// @dynatrace/strato-components/esm/layouts/page-layout/details-layout/DetailsLayout.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { clsx } from "clsx";
import {
  forwardRef,
  memo,
  useContext,
  useMemo,
  useRef
} from "react";
import { useIntl } from "react-intl";
import * as styles from "./DetailsLayout.sty.js";
import { DetailsLayoutContext } from "./DetailsLayoutContext.js";
import { Overlay, OVERLAY_HANDLE_WIDTH } from "./Overlay.js";
import { useLayoutState } from "./useLayoutState.js";
import { useMergeRefs } from "../../../core/hooks/useMergeRefs.js";
import { useSafeLayoutEffect } from "../../../core/hooks/useSafeLayoutEffect.js";
import { Drawer } from "../../../overlays/drawer/Drawer.js";
import { SplitLayout } from "../../split-layout/SplitLayout.js";
import { convertToPx } from "../../split-layout/width-utils.js";
import {
  PageLayoutPortalContext
} from "../shared/PageLayoutPortalContext.js";
import { useTeleport } from "../shared/useTeleport.js";
function DetailsLayoutInner({
  details,
  content,
  containerRef,
  persistedPanelWidthRef,
  contentTeleport,
  panelTeleport
}) {
  const intl = useIntl();
  const layout = useLayoutState({ details, containerRef });
  const detailsAriaLabel = intl.formatMessage({
    defaultMessage: "Details",
    id: "o3A99RFWS+N7bJe8",
    description: "Accessible label for the details panel."
  });
  const contextValue = useMemo(
    () => ({
      collapsed: layout.collapsed,
      setCollapsed: layout.setCollapsed,
      preferredLayout: layout.preferredLayoutMode,
      setPreferredLayout: layout.setPreferredLayoutMode,
      mode: layout.mode
    }),
    [
      layout.collapsed,
      layout.setCollapsed,
      layout.preferredLayoutMode,
      layout.setPreferredLayoutMode,
      layout.mode
    ]
  );
  const pageLayoutPortalContext = useContext(PageLayoutPortalContext);
  useSafeLayoutEffect(() => {
    if (!pageLayoutPortalContext) {
      return;
    }
    pageLayoutPortalContext.registry.register(
      "detailsLayoutContext",
      contextValue
    );
  }, [pageLayoutPortalContext, contextValue]);
  const contentNode = /* @__PURE__ */ jsx(
    "div",
    {
      ref: content.ref,
      ...content.domProps,
      className: clsx(styles.content, content.domProps.className),
      children: /* @__PURE__ */ jsx(contentTeleport.Target, {})
    }
  );
  const layoutNode = (() => {
    switch (layout.mode) {
      case "collapsed": {
        return contentNode;
      }
      case "split": {
        return /* @__PURE__ */ jsxs(SplitLayout, { children: [
          /* @__PURE__ */ jsx(
            SplitLayout.Content,
            {
              ...content.domProps,
              ref: content.ref,
              minWidth: content.config.minWidth,
              className: clsx(styles.content, content.domProps.className),
              children: /* @__PURE__ */ jsx(contentTeleport.Target, {})
            }
          ),
          /* @__PURE__ */ jsx(
            SplitLayout.Panel,
            {
              ...details.domProps,
              ref: details.ref,
              minWidth: details.config.minWidth,
              maxWidth: details.config.maxWidth,
              defaultWidth: persistedPanelWidthRef.current ?? details.config.defaultWidth,
              onResize: (value) => {
                persistedPanelWidthRef.current = value;
                details.config.onResize?.(value);
              },
              resizable: details.config.resizable,
              position: "end",
              className: clsx(styles.panel, details.domProps.className),
              "aria-label": detailsAriaLabel,
              children: /* @__PURE__ */ jsx(panelTeleport.Target, {})
            }
          )
        ] });
      }
      case "overlay": {
        if (layout.containerWidth === 0) {
          return /* @__PURE__ */ jsxs(Fragment, { children: [
            contentNode,
            /* @__PURE__ */ jsx("div", { hidden: true, "aria-hidden": "true", children: /* @__PURE__ */ jsx(panelTeleport.Target, {}) })
          ] });
        }
        return /* @__PURE__ */ jsxs(Fragment, { children: [
          contentNode,
          /* @__PURE__ */ jsx(
            Overlay,
            {
              ...details.domProps,
              ref: details.ref,
              minWidth: convertToPx(
                details.config.minWidth,
                layout.containerWidth
              ),
              maxWidth: details.config.resizable ? layout.containerWidth - OVERLAY_HANDLE_WIDTH : layout.containerWidth,
              defaultWidth: persistedPanelWidthRef.current ?? convertToPx(details.config.defaultWidth, layout.containerWidth),
              resizable: details.config.resizable,
              onResize: (value) => {
                persistedPanelWidthRef.current = value;
                details.config.onResize?.(value);
              },
              children: /* @__PURE__ */ jsx(panelTeleport.Target, {})
            }
          )
        ] });
      }
      case "drawer": {
        return /* @__PURE__ */ jsxs(Fragment, { children: [
          contentNode,
          /* @__PURE__ */ jsx(
            Drawer,
            {
              ...details.domProps,
              ref: details.ref,
              className: clsx(styles.drawer, details.domProps.className),
              "aria-label": detailsAriaLabel,
              isDismissed: layout.collapsed,
              onDismiss: (state) => layout.setCollapsed(state),
              placement: "right",
              width: "wide",
              children: /* @__PURE__ */ jsx("div", { className: styles.drawerContent, children: /* @__PURE__ */ jsx(panelTeleport.Target, {}) })
            }
          )
        ] });
      }
    }
  })();
  return /* @__PURE__ */ jsxs(DetailsLayoutContext.Provider, { value: contextValue, children: [
    /* @__PURE__ */ jsx(panelTeleport.Source, { children: details.children }),
    layoutNode
  ] });
}
const DetailsLayout = memo(
  forwardRef((props, ref) => {
    const { className: rootClassName, details, content, ...rootProps } = props;
    const containerRef = useRef(null);
    const rootRef = useMergeRefs([ref, containerRef]);
    const persistedPanelWidth = useRef();
    const contentTeleport = useTeleport();
    const panelTeleport = useTeleport();
    return /* @__PURE__ */ jsxs(Fragment, { children: [
      /* @__PURE__ */ jsx(contentTeleport.Source, { children: content.children }),
      /* @__PURE__ */ jsx(
        "div",
        {
          ...rootProps,
          ref: rootRef,
          className: clsx(styles.layoutRoot, rootClassName),
          children: details ? /* @__PURE__ */ jsx(
            DetailsLayoutInner,
            {
              details,
              content,
              containerRef,
              persistedPanelWidthRef: persistedPanelWidth,
              contentTeleport,
              panelTeleport
            }
          ) : /* @__PURE__ */ jsx(
            "div",
            {
              ref: content.ref,
              ...content.domProps,
              className: clsx(styles.content, content.domProps.className),
              children: /* @__PURE__ */ jsx(contentTeleport.Target, {})
            }
          )
        }
      )
    ] });
  })
);
DetailsLayout.displayName = "DetailsLayout (forwardRef)";
export {
  DetailsLayout
};
