// @dynatrace/strato-components/esm/layouts/page-layout/details-layout/DetailsControlBar.js
import { jsx, jsxs } from "react/jsx-runtime";
import { clsx } from "clsx";
import { Children } from "react";
import { useIntl } from "react-intl";
import {
  OverlayRightIcon,
  SplitPanelRightIcon,
  XmarkIcon
} from "@dynatrace/strato-icons";
import * as styles from "./DetailsControlBar.sty.js";
import { useDetailsLayoutContext } from "./DetailsLayoutContext.js";
import { Button } from "../../../buttons/button/Button.js";
import { Divider } from "../../divider/Divider.js";
const DetailsControlBar = (props) => {
  const { children, className, ...restProps } = props;
  const context = useDetailsLayoutContext();
  const intl = useIntl();
  const hasCustomActions = Children.count(children) > 0;
  const closeLabel = intl.formatMessage({
    defaultMessage: "Close details",
    id: "UK0OoP7XF++evklW",
    description: "Label for the button that closes the details panel."
  });
  const switchToOverlayLabel = intl.formatMessage({
    defaultMessage: "Switch to overlay",
    id: "+mEjL2q8RACwzIMR",
    description: "Label for the button that switches the details panel to overlay mode."
  });
  const switchToSplitLabel = intl.formatMessage({
    defaultMessage: "Switch to side-by-side",
    id: "RST92XLweEz0Qott",
    description: "Label for the button that switches the details panel to split mode."
  });
  const handleClose = () => {
    context?.setCollapsed(true);
  };
  const handleLayoutToggle = () => {
    if (!context) {
      return;
    }
    context.setPreferredLayout(
      (prev) => prev === "split" ? "overlay" : "split"
    );
  };
  const showLayoutToggle = context?.mode ? ["overlay", "split"].includes(context.mode) : false;
  const isSplit = context?.preferredLayout === "split";
  return /* @__PURE__ */ jsxs("div", { ...restProps, className: clsx(styles.controlBar, className), children: [
    hasCustomActions && /* @__PURE__ */ jsx("div", { className: styles.customActions, children }),
    hasCustomActions && /* @__PURE__ */ jsx(Divider, { orientation: "vertical", as: "hr", flexItem: true }),
    /* @__PURE__ */ jsxs("div", { className: styles.defaultActions, children: [
      showLayoutToggle && /* @__PURE__ */ jsx(
        Button,
        {
          "aria-label": isSplit ? switchToOverlayLabel : switchToSplitLabel,
          onClick: handleLayoutToggle,
          children: /* @__PURE__ */ jsx(Button.Prefix, { children: isSplit ? /* @__PURE__ */ jsx(OverlayRightIcon, {}) : /* @__PURE__ */ jsx(SplitPanelRightIcon, {}) })
        }
      ),
      /* @__PURE__ */ jsx(
        Button,
        {
          "aria-label": closeLabel,
          onClick: handleClose,
          "data-dt-name": "CloseX",
          children: /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(XmarkIcon, {}) })
        }
      )
    ] })
  ] });
};
DetailsControlBar.displayName = "PageLayout.Details.ControlBar";
export {
  DetailsControlBar
};
