// @dynatrace/strato-components/esm/layouts/page-layout/details-layout/Overlay.js
import { jsx, jsxs } from "react/jsx-runtime";
import { useGesture } from "@use-gesture/react";
import { clsx } from "clsx";
import {
  forwardRef,
  memo,
  useRef,
  useState
} from "react";
import * as styles from "./Overlay.sty.js";
import { useFocusRing } from "../../../core/styles/useFocusRing.js";
import { clamp } from "../../split-layout/width-utils.js";
const OVERLAY_HANDLE_WIDTH = 8;
const Overlay = memo(
  forwardRef((props, forwardedRef) => {
    const {
      children,
      className,
      style,
      minWidth = 200,
      maxWidth = Infinity,
      defaultWidth = 300,
      resizable = true,
      onResize,
      ...restProps
    } = props;
    const containerRef = useRef(null);
    const handleRef = useRef(null);
    const { focusProps, focusClassName } = useFocusRing({
      focusWithin: true,
      ignoreModality: true
    });
    const dragStartWidth = useRef(null);
    const clampContentWidth = (width) => clamp(minWidth, width, maxWidth);
    const clampedDefaultWidth = clampContentWidth(defaultWidth);
    const [resizedWidth, setResizedWidth] = useState(null);
    const contentWidth = resizedWidth === null ? clampedDefaultWidth : clampContentWidth(resizedWidth);
    const overlayWidth = resizable ? contentWidth + OVERLAY_HANDLE_WIDTH : contentWidth;
    useGesture(
      {
        onDragStart: () => {
          dragStartWidth.current = clampContentWidth(contentWidth);
        },
        onDragEnd: () => {
          dragStartWidth.current = null;
        },
        onDrag: (state) => {
          if (dragStartWidth.current === null) {
            return;
          }
          const nextWidth = clamp(
            minWidth,
            dragStartWidth.current + state.movement[0] * -1,
            maxWidth
          );
          setResizedWidth(nextWidth);
          onResize?.(nextWidth);
        }
      },
      { target: handleRef }
    );
    return /* @__PURE__ */ jsxs(
      "div",
      {
        role: "dialog",
        className: styles.overlay,
        style: { ...style, width: overlayWidth },
        ref: containerRef,
        children: [
          resizable ? /* @__PURE__ */ jsx("div", { className: styles.handleContainer, children: /* @__PURE__ */ jsx(
            "hr",
            {
              ...focusProps,
              ref: handleRef,
              className: clsx(styles.resizeHandle, focusClassName),
              tabIndex: 0
            }
          ) }) : null,
          /* @__PURE__ */ jsx(
            "div",
            {
              ...restProps,
              ref: forwardedRef,
              className: clsx(styles.overlayContent, className),
              children
            }
          )
        ]
      }
    );
  })
);
Overlay.displayName = "Overlay";
export {
  OVERLAY_HANDLE_WIDTH,
  Overlay
};
