// @dynatrace/strato-components/esm/editors/code-editor/languages/languages.js
import { javascript } from "@codemirror/lang-javascript";
import { json, jsonParseLinter } from "@codemirror/lang-json";
import { markdown } from "@codemirror/lang-markdown";
import { StreamLanguage } from "@codemirror/language";
import { yaml } from "@codemirror/legacy-modes/mode/yaml";
import { linter } from "@codemirror/lint";
import { javascriptAutocompleteProvider } from "../auto-completion/javascript.js";
const codeEditorLanguages = {
  other: {
    extendWith: []
  },
  json: {
    language: json().language,
    extendWith: [linter(jsonParseLinter())]
  },
  js: {
    language: javascript().language,
    extendWith: [],
    autocompleteOptions: {
      activateOnTyping: true,
      providers: [javascriptAutocompleteProvider]
    }
  },
  jsx: {
    language: javascript({ jsx: true }).language,
    extendWith: [],
    autocompleteOptions: {
      activateOnTyping: true,
      providers: [javascriptAutocompleteProvider]
    }
  },
  ts: {
    language: javascript({ typescript: true }).language,
    extendWith: [],
    autocompleteOptions: {
      activateOnTyping: true,
      providers: [javascriptAutocompleteProvider]
    }
  },
  tsx: {
    language: javascript({ typescript: true, jsx: true }).language,
    extendWith: [],
    autocompleteOptions: {
      activateOnTyping: true,
      providers: [javascriptAutocompleteProvider]
    }
  },
  md: {
    language: markdown().language,
    extendWith: []
  },
  yaml: {
    language: StreamLanguage.define(yaml),
    extendWith: []
  }
};
const CodeEditorLanguageKeys = Object.keys(
  codeEditorLanguages
);
export {
  CodeEditorLanguageKeys,
  codeEditorLanguages
};
