// @dynatrace/strato-components/esm/content/accordion/Accordion.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { forwardRef, useRef, Children } from "react";
import { AccordionActions, useAccordionState } from "./accordion-state.js";
import { accordionDividerMarginCSS } from "./Accordion.sty.js";
import { AccordionSection } from "./AccordionSection.js";
import { AccordionSectionContent } from "./AccordionSectionContent.js";
import { AccordionSectionLabel } from "./AccordionSectionLabel.js";
import { AccordionContextProvider } from "./contexts/AccordionContext.js";
import { useAriaLabelingProps } from "../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useSafeForwardProps } from "../../core/hooks/useSafeForwardProps.js";
import { Divider } from "../../layouts/divider/Divider.js";
import { Flex } from "../../layouts/flex/Flex.js";
const COMPONENT_NAME = "Accordion";
function RootAccordion(props, forwardedRef) {
  const {
    expanded,
    defaultExpanded,
    onExpandChange,
    multiple = false,
    triggerPosition = "end",
    interactive = true,
    color = "neutral",
    showDividers = true,
    size = "default",
    children,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
  const ariaLabelingProps = useAriaLabelingProps(restProps);
  const forwardProps = useSafeForwardProps(restProps, ariaLabelingProps);
  const sectionCount = Children.count(children);
  const accordionSections = Children.map(children, (child, index) => /* @__PURE__ */ jsxs(Fragment, { children: [
    child,
    showDividers && sectionCount > 1 && index < sectionCount - 1 && /* @__PURE__ */ jsx(Divider, { className: accordionDividerMarginCSS[size], color })
  ] }));
  const [state, dispatch] = useAccordionState({
    expanded,
    defaultExpanded,
    multiple,
    onExpandChange
  });
  const lastMultiple = useRef(multiple);
  if (multiple !== lastMultiple.current) {
    dispatch({ type: AccordionActions.SET_MULTIPLE, payload: multiple });
    lastMultiple.current = multiple;
  }
  return /* @__PURE__ */ jsx(
    "div",
    {
      ref: forwardedRef,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...ariaLabelingProps,
      ...behavioralTrackingProps,
      ...forwardProps,
      "data-dt-component": COMPONENT_NAME,
      children: /* @__PURE__ */ jsx(
        AccordionContextProvider,
        {
          value: {
            triggerPosition,
            size,
            interactive,
            state,
            color,
            dispatch
          },
          children: /* @__PURE__ */ jsx(
            Flex,
            {
              flexDirection: "column",
              gap: showDividers ? 0 : size === "default" ? 8 : 4,
              children: accordionSections
            }
          )
        }
      )
    }
  );
}
RootAccordion.displayName = COMPONENT_NAME;
const AccordionForwardRef = forwardRef(RootAccordion);
const Accordion = Object.assign(AccordionForwardRef, {
  Section: AccordionSection,
  SectionLabel: AccordionSectionLabel,
  SectionContent: AccordionSectionContent
});
export {
  Accordion
};
