// @dynatrace/strato-components/esm/content/accordion/AccordionSection.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import {
  forwardRef,
  useCallback,
  useLayoutEffect,
  useRef,
  useState
} from "react";
import useResizeObserver from "use-resize-observer";
import { ChevronDownSmallIcon } from "@dynatrace/strato-icons";
import { AccordionActions } from "./accordion-state.js";
import {
  accordionTriggerIconCSS,
  accordionTriggerIconEndCSS,
  accordionTriggerIconExpandedCSS,
  accordionTriggerIconStartCSS
} from "./Accordion.sty.js";
import { AccordionSectionContent } from "./AccordionSectionContent.js";
import { AccordionSectionLabel } from "./AccordionSectionLabel.js";
import { useAccordionContext } from "./contexts/AccordionContext.js";
import { AccordionSectionContext } from "./contexts/AccordionSectionContext.js";
import { Button } from "../../buttons/button/Button.js";
import { useId } from "../../core/hooks/useId.js";
import { useSlots } from "../../core/hooks/useSlots.js";
import { Flex } from "../../layouts/flex/Flex.js";
import { fieldTokens as fieldColorsCSS } from "../../styles/field.sty.js";
const AccordionSection = /* @__PURE__ */ forwardRef(
  (props, forwardedRef) => {
    const {
      disabled = false,
      id: idProp,
      sectionId,
      children,
      className: consumerClassName,
      style: consumerStyle,
      keepMounted = false,
      "data-testid": dataTestId,
      as,
      ...remainingProps
    } = props;
    const [slots] = useSlots(children, {
      label: AccordionSectionLabel,
      content: AccordionSectionContent
    });
    const { triggerPosition, interactive, color, size, state, dispatch } = useAccordionContext();
    const sectionColor = props.color ?? color;
    const triggerId = useId();
    const contentId = useId();
    const triggerIsStart = triggerPosition === "start";
    const id = sectionId ?? idProp;
    const isExpanded = state.expanded.has(id);
    const labelId = useId();
    const containerRef = useRef(null);
    const { width } = useResizeObserver({ ref: containerRef.current });
    const [offsetWidth, setOffsetWidth] = useState(0);
    const updateOverflowWidth = useCallback(() => {
      if (containerRef.current) {
        setOffsetWidth(
          containerRef.current.scrollWidth - containerRef.current.offsetWidth
        );
      }
    }, []);
    const dataDtProperties = `state:${isExpanded ? "collapse" : "expand"}`;
    useLayoutEffect(() => {
      updateOverflowWidth();
    }, [updateOverflowWidth, width]);
    const trigger = /* @__PURE__ */ jsx(
      Button,
      {
        id: triggerId,
        onClick: () => {
          if (!interactive) {
            dispatch({ type: AccordionActions.TOGGLE_SECTION, payload: id });
          }
        },
        color: sectionColor,
        size,
        disabled,
        "aria-labelledby": labelId,
        "aria-expanded": isExpanded,
        "aria-controls": isExpanded || keepMounted ? contentId : void 0,
        "data-dt-name": "ExpandCollapse",
        "data-dt-properties": dataDtProperties,
        children: /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(
          ChevronDownSmallIcon,
          {
            className: clsx(accordionTriggerIconCSS, {
              [accordionTriggerIconStartCSS]: triggerIsStart,
              [accordionTriggerIconEndCSS]: !triggerIsStart,
              [accordionTriggerIconExpandedCSS]: isExpanded
            })
          }
        ) })
      }
    );
    const Section = as ?? "div";
    return /* @__PURE__ */ jsx(
      Section,
      {
        "data-testid": dataTestId,
        ref: forwardedRef,
        className: clsx(
          fieldColorsCSS({ color: sectionColor }),
          consumerClassName
        ),
        style: consumerStyle,
        ...remainingProps,
        children: (slots.label || slots.content) && /* @__PURE__ */ jsxs(
          AccordionSectionContext.Provider,
          {
            value: {
              labelId,
              contentId,
              isExpanded,
              keepMounted,
              size
            },
            children: [
              /* @__PURE__ */ jsxs(
                Flex,
                {
                  ref: containerRef,
                  flexDirection: "row",
                  alignItems: "center",
                  justifyContent: triggerIsStart ? void 0 : "space-between",
                  gap: size === "default" ? 8 : 4,
                  id: labelId,
                  onClick: () => {
                    if (interactive && !disabled) {
                      dispatch({
                        type: AccordionActions.TOGGLE_SECTION,
                        payload: id
                      });
                    }
                  },
                  style: {
                    cursor: interactive && !disabled ? "pointer" : void 0,
                    width: `calc(100% - ${offsetWidth}px)`
                  },
                  "data-dt-name": "ExpandCollapseHeader",
                  "data-dt-properties": dataDtProperties,
                  children: [
                    triggerIsStart && trigger,
                    slots.label,
                    !triggerIsStart && trigger
                  ]
                }
              ),
              slots.content
            ]
          }
        )
      }
    );
  }
);
AccordionSection.displayName = "Accordion.Section";
export {
  AccordionSection
};
