// @dynatrace/strato-components/esm/content/accordion/Accordion.sty.js
import "../../core/components/app-root/AppRoot.css";
import "../../styles/field.css";
import "./Accordion.css";
import { createRuntimeFn as _7a468 } from "@vanilla-extract/recipes/createRuntimeFn";
var accordionContentCSS = _7a468({ defaultClassName: "_12zc70k8-3-6-0", variantClassNames: { size: { "default": "_12zc70k9-3-6-0", condensed: "_12zc70ka-3-6-0" } }, defaultVariants: {}, compoundVariants: [] });
var accordionContentWrapperCSS = "_12zc70k7-3-6-0";
var accordionDividerMarginCSS = { "default": "_12zc70k1-3-6-0", condensed: "_12zc70k2-3-6-0" };
var accordionLabelCSS = "_12zc70k0-3-6-0";
var accordionTriggerIconCSS = "_12zc70k3-3-6-0";
var accordionTriggerIconEndCSS = "_12zc70k5-3-6-0";
var accordionTriggerIconExpandedCSS = "_12zc70k6-3-6-0";
var accordionTriggerIconStartCSS = "_12zc70k4-3-6-0";
export {
  accordionContentCSS,
  accordionContentWrapperCSS,
  accordionDividerMarginCSS,
  accordionLabelCSS,
  accordionTriggerIconCSS,
  accordionTriggerIconEndCSS,
  accordionTriggerIconExpandedCSS,
  accordionTriggerIconStartCSS
};
