// @dynatrace/strato-components/esm/content/accordion/AccordionSectionContent.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import {
  AnimatePresence,
  motion,
  useReducedMotion
} from "framer-motion";
import { forwardRef, useContext, useState } from "react";
import AnimationsJS from "@dynatrace/strato-design-tokens/animations-js";
import {
  accordionContentCSS,
  accordionContentWrapperCSS
} from "./Accordion.sty.js";
import { AccordionSectionContext } from "./contexts/AccordionSectionContext.js";
import { canReceiveDefaultTextFormat as _canReceiveDefaultTextFormat } from "../../core/utils/can-receive-default-text-format.js";
import { expandCollapseEasing } from "../../core/utils/framer-motion-easings.js";
import { Text } from "../../typography/text/Text.js";
const animationVariantSectionContent = {
  expanded: {
    height: "auto",
    opacity: 1
  },
  collapsed: {
    height: 0,
    opacity: 0
  }
};
const AccordionSectionContent = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const {
    children,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId
  } = props;
  const { labelId, contentId, isExpanded, keepMounted, size } = useContext(
    AccordionSectionContext
  );
  const childIsText = _canReceiveDefaultTextFormat(children);
  const [isAnimating, setIsAnimating] = useState(false);
  const shouldReduceMotion = useReducedMotion();
  const content = /* @__PURE__ */ jsx(
    motion.div,
    {
      "data-testid": dataTestId,
      variants: animationVariantSectionContent,
      initial: "collapsed",
      animate: isExpanded ? "expanded" : "collapsed",
      exit: "collapsed",
      transition: {
        duration: shouldReduceMotion ? 0 : AnimationsJS.ExpandCollapse.Duration.Base,
        ease: expandCollapseEasing
      },
      className: clsx(accordionContentWrapperCSS, consumerClassName),
      style: {
        ...consumerStyle,
        willChange: "initial",
        overflow: keepMounted && (isAnimating || !isExpanded) ? "hidden" : void 0
      },
      onAnimationStart: () => {
        setIsAnimating(true);
      },
      onAnimationComplete: () => {
        setIsAnimating(false);
      },
      children: /* @__PURE__ */ jsx(
        "div",
        {
          className: clsx([
            accordionContentCSS({
              size
            }),
            consumerClassName
          ]),
          ref: forwardedRef,
          id: contentId,
          role: "region",
          "aria-labelledby": labelId,
          children: childIsText ? /* @__PURE__ */ jsx(Text, { textStyle: "base", as: "p", children }) : children
        }
      )
    }
  );
  if (keepMounted) {
    return content;
  }
  return /* @__PURE__ */ jsx("div", { style: { overflow: isAnimating ? "hidden" : void 0 }, children: /* @__PURE__ */ jsx(AnimatePresence, { initial: false, children: isExpanded && content }) });
});
AccordionSectionContent.displayName = "Accordion.SectionContent";
export {
  AccordionSectionContent
};
