// @dynatrace/strato-components/charts/core/components/interactions-state-helpers/reducers/explore-reducers.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var explore_reducers_exports = {};
__export(explore_reducers_exports, {
  reduceExploreCancelAll: () => reduceExploreCancelAll,
  reduceExploreCancelSelect: () => reduceExploreCancelSelect,
  reduceExploreCancelTooltip: () => reduceExploreCancelTooltip,
  reduceExploreEnd: () => reduceExploreEnd,
  reduceExplorePin: () => reduceExplorePin,
  reduceExplorePosition: () => reduceExplorePosition,
  reduceExploreReset: () => reduceExploreReset,
  reduceExploreSelect: () => reduceExploreSelect,
  reduceExploreUnPin: () => reduceExploreUnPin
});
module.exports = __toCommonJS(explore_reducers_exports);
function reduceExplorePosition(position) {
  return (prev) => ({
    ...prev,
    explore: {
      ...prev.explore,
      visible: true,
      position
    }
  });
}
function reduceExploreEnd(prev) {
  return {
    ...prev,
    explore: {
      ...prev.explore,
      position: prev.explore.pinned ? prev.explore.position : void 0,
      visible: prev.explore.pinned
    }
  };
}
function reduceExploreReset(prev) {
  return {
    ...prev,
    highlightedSeriesId: void 0,
    explore: {
      ...prev.explore,
      visible: false,
      pinned: false,
      position: void 0
    },
    select: void 0
  };
}
function reduceExplorePin(position, highlightedSeriesId) {
  return (prev) => ({
    ...prev,
    highlightedSeriesId,
    explore: {
      ...prev.explore,
      pinned: true,
      position
    }
  });
}
function reduceExploreUnPin(position) {
  return (prev) => ({
    ...prev,
    select: void 0,
    highlightedSeriesId: void 0,
    explore: {
      ...prev.explore,
      pinned: false,
      position
    }
  });
}
function reduceExploreSelect(select) {
  return (prev) => ({
    ...prev,
    select: {
      ...prev.select,
      ...select
    },
    explore: {
      visible: false,
      pinned: false
    },
    highlightedSeriesId: void 0
  });
}
function reduceExploreCancelTooltip(prev) {
  return {
    ...prev,
    explore: {
      visible: false,
      pinned: false
    },
    highlightedSeriesId: void 0
  };
}
function reduceExploreCancelSelect(prev) {
  return {
    ...prev,
    select: void 0
  };
}
function reduceExploreCancelAll(prev) {
  return {
    ...prev,
    explore: {
      visible: false,
      pinned: false
    },
    select: void 0,
    highlightedSeriesId: void 0
  };
}
