// @dynatrace/strato-components/charts/core/components/interactions-state-helpers/reducers/reset-reducers.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var reset_reducers_exports = {};
__export(reset_reducers_exports, {
  reduceResetExploreVisible: () => reduceResetExploreVisible,
  reduceResetInteractions: () => reduceResetInteractions,
  reduceResetToInfiniteZoom: () => reduceResetToInfiniteZoom,
  reduceResetZoom: () => reduceResetZoom
});
module.exports = __toCommonJS(reset_reducers_exports);
function reduceResetInteractions(isZoomOutEnabled, position) {
  return (prev) => ({
    ...prev,
    highlightedSeriesId: void 0,
    explore: {
      position,
      visible: position !== void 0,
      pinned: false
    },
    select: void 0,
    currentDomain: void 0,
    zoom: {
      zoomInEnabled: true,
      zoomOutEnabled: isZoomOutEnabled
    }
  });
}
function reduceResetZoom(prev) {
  return {
    ...prev,
    currentDomain: void 0,
    highlightedSeriesId: prev.explore.pinned ? void 0 : prev.highlightedSeriesId,
    zoom: {
      zoomOutEnabled: false,
      zoomInEnabled: true
    }
  };
}
function reduceResetToInfiniteZoom(prev) {
  return {
    ...prev,
    currentDomain: void 0,
    highlightedSeriesId: void 0,
    zoom: {
      zoomOutEnabled: true,
      zoomInEnabled: true
    }
  };
}
function reduceResetExploreVisible(prev) {
  return {
    ...prev,
    explore: {
      ...prev.explore,
      visible: false
    }
  };
}
