// @dynatrace/strato-components/charts/core/components/copy-to-clipboard/copy-to-clipboard-messages.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var copy_to_clipboard_messages_exports = {};
__export(copy_to_clipboard_messages_exports, {
  COPY_TO_CLIPBOARD_MESSAGES: () => COPY_TO_CLIPBOARD_MESSAGES
});
module.exports = __toCommonJS(copy_to_clipboard_messages_exports);
var import_react_intl = require("react-intl");
const COPY_TO_CLIPBOARD_MESSAGES = (0, import_react_intl.defineMessages)({
  copyActionName: {
    id: "6vA3zfYytiSPVBZV",
    defaultMessage: "Copy name",
    description: "Text in a tooltip shown when clicking a tooltip item and hovering the copy button next to its name"
  },
  copyCoordinates: {
    id: "Rrlag1wRQFMpYHX7",
    defaultMessage: "Copy coordinates",
    description: "Text in a tooltip shown when clicking a geo map tooltip item and hovering the copy button next to its coordinates"
  },
  tooltipActionCopy: {
    id: "FxsZTOljj8LWcZO0",
    defaultMessage: "Copy",
    description: "Text in a tooltip shown by default when hovering"
  },
  actionCopied: {
    id: "aYOFXw5HKPdKKWSu",
    defaultMessage: "Copied",
    description: "Text in a tooltip shown when clicking a tooltip series and copying its series name"
  },
  copyField: {
    id: "XOE5zKTRmqYLBnRf",
    defaultMessage: "Copy {field}",
    description: "Label for the copy action in the tooltip, where field is the name of the field to be copied"
  }
});
