// @dynatrace/strato-components/esm/filters/segment-selector/parts/noAccessRights/NoAccessWithoutSelection.js
import { jsx, jsxs } from "react/jsx-runtime";
import { useEffect, useRef } from "react";
import { useIntl } from "react-intl";
import { ExternalLinkIcon } from "@dynatrace/strato-icons";
import { Button } from "../../../../buttons/button/Button.js";
import { SEGMENTS_DOCUMENTATION_URL } from "../../../../core/external-links.js";
import { Flex } from "../../../../layouts/flex/Flex.js";
import { Text } from "../../../../typography/text/Text.js";
import { SEGMENTSELECTOR_MESSAGES } from "../../segmentMessages.js";
import * as styles from "../Overlay.sty.js";
import { CopyPermissionsButton } from "./CopyPermissionsButton.js";
const NoAccessWithoutSelection = () => {
  const intl = useIntl();
  const buttonRef = useRef(null);
  useEffect(() => {
    buttonRef.current?.focus();
  }, []);
  return /* @__PURE__ */ jsxs(
    Flex,
    {
      flexDirection: "column",
      alignItems: "center",
      padding: 16,
      width: "100vw",
      className: styles.overlayDefaultWidth,
      style: { maxWidth: "320px", textAlign: "center" },
      children: [
        /* @__PURE__ */ jsx(Text, { textStyle: "base-emphasized", children: intl.formatMessage(SEGMENTSELECTOR_MESSAGES.noAccessNoSegments) }),
        intl.formatMessage(SEGMENTSELECTOR_MESSAGES.copyDetailsHint),
        /* @__PURE__ */ jsxs(Flex, { children: [
          /* @__PURE__ */ jsx(CopyPermissionsButton, { variant: "emphasized", ref: buttonRef }),
          /* @__PURE__ */ jsxs(
            Button,
            {
              as: "a",
              href: SEGMENTS_DOCUMENTATION_URL,
              target: "_blank",
              rel: "noopener noreferrer",
              children: [
                intl.formatMessage(SEGMENTSELECTOR_MESSAGES.docsLinkTextShort),
                /* @__PURE__ */ jsx(Button.Suffix, { children: /* @__PURE__ */ jsx(ExternalLinkIcon, {}) })
              ]
            }
          )
        ] })
      ]
    }
  );
};
export {
  NoAccessWithoutSelection
};
