// @dynatrace/strato-components/charts/core/constants.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var constants_exports = {};
__export(constants_exports, {
  CATEGORICAL_BARS_INNER_PADDING: () => CATEGORICAL_BARS_INNER_PADDING,
  CATEGORICAL_BARS_OUTER_PADDING: () => CATEGORICAL_BARS_OUTER_PADDING,
  CSS_DOWNLOAD_FILENAME: () => CSS_DOWNLOAD_FILENAME,
  ChartOverlaysElevation: () => ChartOverlaysElevation,
  DEFAULT_CUSTOM_TOOLTIP_CONFIG: () => DEFAULT_CUSTOM_TOOLTIP_CONFIG,
  DEFAULT_DATA_POINT_RADIUS: () => DEFAULT_DATA_POINT_RADIUS,
  DEFAULT_FILL_OPACITY: () => DEFAULT_FILL_OPACITY,
  DEFAULT_HIGHLIGHTED_OPACITY: () => DEFAULT_HIGHLIGHTED_OPACITY,
  DEFAULT_INACTIVE_FILL_OPACITY: () => DEFAULT_INACTIVE_FILL_OPACITY,
  DEFAULT_INFINITE_ZOOM: () => DEFAULT_INFINITE_ZOOM,
  DEFAULT_OPACITY_BAND_CHART: () => DEFAULT_OPACITY_BAND_CHART,
  DEFAULT_SELECT_ENABLED: () => DEFAULT_SELECT_ENABLED,
  DEFAULT_SERIES_CURVE_SHAPE: () => DEFAULT_SERIES_CURVE_SHAPE,
  DEFAULT_STROKE_WIDTH_BAND_CHART: () => DEFAULT_STROKE_WIDTH_BAND_CHART,
  DEFAULT_TRUNCATION_MODE: () => DEFAULT_TRUNCATION_MODE,
  DEFAULT_VALUE_ACCESSOR: () => DEFAULT_VALUE_ACCESSOR,
  DEFAULT_VALUE_REPRESENTATION: () => DEFAULT_VALUE_REPRESENTATION,
  FONT_BASE: () => FONT_BASE,
  GRAIL_MAX_DOMAIN: () => GRAIL_MAX_DOMAIN,
  GRAIL_MIN_DOMAIN: () => GRAIL_MIN_DOMAIN,
  INACTIVE_FILL_OPACITY_AREA_CHART: () => INACTIVE_FILL_OPACITY_AREA_CHART,
  INACTIVE_FILL_OPACITY_BAND_CHART: () => INACTIVE_FILL_OPACITY_BAND_CHART,
  INACTIVE_FILL_OPACITY_BAR_CHART: () => INACTIVE_FILL_OPACITY_BAR_CHART,
  INACTIVE_FILL_OPACITY_LINE_CHART: () => INACTIVE_FILL_OPACITY_LINE_CHART,
  INACTIVE_FILL_OPACITY_PIE_CHART: () => INACTIVE_FILL_OPACITY_PIE_CHART,
  INACTIVE_FILL_OPACITY_PIE_CHART_LABEL: () => INACTIVE_FILL_OPACITY_PIE_CHART_LABEL,
  INACTIVE_STROKE_ONLY_OPACITY_BAND_CHART: () => INACTIVE_STROKE_ONLY_OPACITY_BAND_CHART,
  LAYOUT_PADDING: () => LAYOUT_PADDING,
  LEGEND_POSITION: () => LEGEND_POSITION,
  LOADING_STATE_OPACITY: () => LOADING_STATE_OPACITY,
  MIDDLE_BUTTON_ID: () => MIDDLE_BUTTON_ID,
  ORIENTATION: () => ORIENTATION,
  PINCH_RELATIVE_THRESHOLD: () => PINCH_RELATIVE_THRESHOLD,
  RIGHT_BUTTON_ID: () => RIGHT_BUTTON_ID,
  SCALE_FALLBACK_COLOR: () => SCALE_FALLBACK_COLOR,
  SIXTY_FPS: () => SIXTY_FPS,
  STROKE_ONLY_OPACITY_BAND_CHART: () => STROKE_ONLY_OPACITY_BAND_CHART,
  SUFFIX_RATIO: () => SUFFIX_RATIO,
  SWIPE_RELATIVE_THRESHOLD: () => SWIPE_RELATIVE_THRESHOLD,
  TRUNCATION_MODE: () => TRUNCATION_MODE,
  defaultRatio: () => defaultRatio
});
module.exports = __toCommonJS(constants_exports);
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
const SCALE_FALLBACK_COLOR = import_colors.default.Background.Container.Neutral.Accent;
const SIXTY_FPS = 1e3 / 60;
const DEFAULT_DATA_POINT_RADIUS = 3;
const DEFAULT_FILL_OPACITY = 0.8;
const DEFAULT_HIGHLIGHTED_OPACITY = 0.7;
const DEFAULT_INACTIVE_FILL_OPACITY = 0.1;
const INACTIVE_FILL_OPACITY_BAR_CHART = DEFAULT_INACTIVE_FILL_OPACITY;
const INACTIVE_FILL_OPACITY_AREA_CHART = DEFAULT_INACTIVE_FILL_OPACITY;
const INACTIVE_FILL_OPACITY_LINE_CHART = DEFAULT_INACTIVE_FILL_OPACITY;
const INACTIVE_FILL_OPACITY_PIE_CHART = DEFAULT_INACTIVE_FILL_OPACITY;
const INACTIVE_FILL_OPACITY_PIE_CHART_LABEL = 0.2;
const DEFAULT_OPACITY_BAND_CHART = 0.5;
const INACTIVE_FILL_OPACITY_BAND_CHART = 0.15;
const STROKE_ONLY_OPACITY_BAND_CHART = 0.8;
const INACTIVE_STROKE_ONLY_OPACITY_BAND_CHART = 0.15;
const DEFAULT_STROKE_WIDTH_BAND_CHART = 1;
const defaultRatio = "auto";
const DEFAULT_TRUNCATION_MODE = "middle";
const DEFAULT_VALUE_REPRESENTATION = "absolute";
const DEFAULT_VALUE_ACCESSOR = "value";
const DEFAULT_SERIES_CURVE_SHAPE = "linear";
const DEFAULT_INFINITE_ZOOM = false;
const DEFAULT_SELECT_ENABLED = false;
const LAYOUT_PADDING = 8;
const CATEGORICAL_BARS_INNER_PADDING = 0.25;
const CATEGORICAL_BARS_OUTER_PADDING = 0.15;
const CSS_DOWNLOAD_FILENAME = "data.csv";
const LEGEND_POSITION = Object.freeze({
  Auto: "auto",
  Bottom: "bottom",
  Right: "right"
});
const TRUNCATION_MODE = Object.freeze({
  End: "end",
  Middle: "middle",
  Start: "start"
});
const GRAIL_MIN_DOMAIN = /* @__PURE__ */ new Date("1677-09-21T00:12:43.145224192Z");
const GRAIL_MAX_DOMAIN = /* @__PURE__ */ new Date("2262-04-11T23:47:16.854775807Z");
const SWIPE_RELATIVE_THRESHOLD = 0.15;
const PINCH_RELATIVE_THRESHOLD = 0.15;
const RIGHT_BUTTON_ID = 2;
const MIDDLE_BUTTON_ID = 4;
const ChartOverlaysElevation = {
  crosshair: 1,
  dataPoints: 2,
  brush: 3,
  toolbar: 4
};
const FONT_BASE = 10;
const SUFFIX_RATIO = 0.66;
const LOADING_STATE_OPACITY = 40;
const ORIENTATION = {
  Horizontal: "horizontal",
  Vertical: "vertical"
};
const DEFAULT_CUSTOM_TOOLTIP_CONFIG = {
  hidden: false
};
