// @dynatrace/strato-components/forms/select/styles/Select.sty.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Select_css_exports = {};
__export(Select_css_exports, {
  clearReactSelectZIndex: () => clearReactSelectZIndex,
  filter: () => filter,
  placeholderColor: () => placeholderColor,
  select: () => select,
  selectAll: () => selectAll,
  selectGroupHeading: () => selectGroupHeading,
  selectMenuDivider: () => selectMenuDivider,
  selectMenuListWrapper: () => selectMenuListWrapper,
  selectMenuLoadingWrapper: () => selectMenuLoadingWrapper,
  selectMenuOverlay: () => selectMenuOverlay,
  selectMultiValue: () => selectMultiValue,
  selectNoOptionsMessage: () => selectNoOptionsMessage,
  selectOptionMulti: () => selectOptionMulti,
  selectOptionMultiIndicator: () => selectOptionMultiIndicator,
  selectOptionMultiIndicatorIcon: () => selectOptionMultiIndicatorIcon,
  selectOptionPrefix: () => selectOptionPrefix,
  selectOptionSingle: () => selectOptionSingle,
  selectOptionSuffix: () => selectOptionSuffix,
  selectTrigger: () => selectTrigger,
  selectTriggerIconWrapper: () => selectTriggerIconWrapper,
  selectValueContainerTruncationBox: () => selectValueContainerTruncationBox,
  selectVirtualizationWrapper: () => selectVirtualizationWrapper,
  selectWrapper: () => selectWrapper,
  validAndEnabled: () => validAndEnabled
});
module.exports = __toCommonJS(Select_css_exports);
var import_AppRoot_css_ts_vanilla = require("../../../core/components/app-root/AppRoot.css");
var import_field_css_ts_vanilla = require("../../../styles/field.css");
var import_Select_css_ts_vanilla = require("./Select.css");
var import_createRuntimeFn = require("@vanilla-extract/recipes/createRuntimeFn");
var clearReactSelectZIndex = "_dlenfc5-3-6-0";
var filter = "_dlenfc1-3-6-0";
var placeholderColor = "_dlenfcl-3-6-0";
var select = "_dlenfc0-3-6-0";
var selectAll = "_dlenfcs-3-6-0";
var selectGroupHeading = "_dlenfcf-3-6-0";
var selectMenuDivider = "_dlenfcn-3-6-0";
var selectMenuListWrapper = "_dlenfch-3-6-0";
var selectMenuLoadingWrapper = "_dlenfcm-3-6-0";
var selectMenuOverlay = (0, import_createRuntimeFn.createRuntimeFn)({ defaultClassName: "_dlenfco-3-6-0", variantClassNames: { isTriggerWidth: { true: "_dlenfcp-3-6-0" } }, defaultVariants: { isTriggerWidth: false }, compoundVariants: [] });
var selectMultiValue = "_dlenfcq-3-6-0";
var selectNoOptionsMessage = "_dlenfci-3-6-0";
var selectOptionMulti = "_dlenfct-3-6-0";
var selectOptionMultiIndicator = (0, import_createRuntimeFn.createRuntimeFn)({ defaultClassName: "_dlenfcu-3-6-0", variantClassNames: { selected: { true: "_dlenfcv-3-6-0" } }, defaultVariants: {}, compoundVariants: [] });
var selectOptionMultiIndicatorIcon = "_dlenfcw-3-6-0";
var selectOptionPrefix = "_dlenfcx-3-6-0";
var selectOptionSingle = "_dlenfcr-3-6-0";
var selectOptionSuffix = "_dlenfcy-3-6-0";
var selectTrigger = (0, import_createRuntimeFn.createRuntimeFn)({ defaultClassName: "_dlenfc9-3-6-0", variantClassNames: { fullWidth: { true: "_dlenfca-3-6-0" }, showError: { true: "_dlenfcb-3-6-0" }, readOnly: { true: "_dlenfcc-3-6-0" } }, defaultVariants: { fullWidth: false }, compoundVariants: [[{ showError: true, readOnly: true }, "_dlenfcd-3-6-0"], [{ showError: false, readOnly: true }, "_dlenfce-3-6-0"]] });
var selectTriggerIconWrapper = (0, import_createRuntimeFn.createRuntimeFn)({ defaultClassName: "_dlenfc6-3-6-0", variantClassNames: { isOverlayOpen: { true: "_dlenfc7-3-6-0" }, readOnly: { true: "_dlenfc8-3-6-0" } }, defaultVariants: { isOverlayOpen: false }, compoundVariants: [] });
var selectValueContainerTruncationBox = "_dlenfcj-3-6-0";
var selectVirtualizationWrapper = "_dlenfcg-3-6-0";
var selectWrapper = (0, import_createRuntimeFn.createRuntimeFn)({ defaultClassName: "_dlenfc2-3-6-0", variantClassNames: { fullWidth: { true: "_dlenfc3-3-6-0", false: "_dlenfc4-3-6-0" } }, defaultVariants: { fullWidth: false }, compoundVariants: [] });
var validAndEnabled = "_dlenfck-3-6-0";
