// @dynatrace/strato-components/esm/charts/tree-map/utils/node-aggregations.js
import { getNodesValueSum } from "./get-nodes-value-sum.js";
import {
  AGGREGATED_NODE_COLOR,
  CLUSTER_OTHER_ID,
  DEFAULT_AGGREGATED_LABEL
} from "../constants.js";
const aggregatedColorToken = {
  Default: AGGREGATED_NODE_COLOR
};
const getAggregatedLeaf = (nodes, clusterName, formatter, clusterColorToken, nameAccessor) => {
  const aggeregatedValue = getNodesValueSum(nodes);
  return {
    id: `other-${clusterName}`,
    value: aggeregatedValue,
    formattedValue: formatter(aggeregatedValue),
    name: DEFAULT_AGGREGATED_LABEL,
    color: nameAccessor ? aggregatedColorToken : clusterColorToken,
    ...nameAccessor && {
      aggregatedNodes: nodes.map((node) => node[nameAccessor])
    }
  };
};
const getAggregatedCluster = (nodes, formatter) => ({
  id: CLUSTER_OTHER_ID,
  value: getNodesValueSum(nodes),
  name: DEFAULT_AGGREGATED_LABEL,
  color: aggregatedColorToken,
  // for highlighting logic, the generated id must be 'other-cluster', same as CLUSTER_OTHER_ID
  children: [
    getAggregatedLeaf(nodes, "cluster", formatter, aggregatedColorToken)
  ],
  aggregatedNodes: nodes.map((node) => node.name)
});
export {
  getAggregatedCluster,
  getAggregatedLeaf
};
