// @dynatrace/strato-components/esm/charts/xy-chart/components/legend/XYLegend/XYLegend.js
import { jsx } from "react/jsx-runtime";
import { useCallback, useMemo } from "react";
import { getXYLegendActions } from "./utils/getXYLegendActions.js";
import {
  AREA_CHART_ICON_SHAPE,
  BAR_CHART_ICON_SHAPE,
  LINE_CHART_ICON_SHAPE,
  RECT_CHART_ICON_SHAPE,
  XYLegendIcon
} from "./XYLegendIcon.js";
import { useChartLayoutLoading } from "../../../../core/components/chart-layout/hooks/useChartLayoutLoading.js";
import { useChartLayoutOrientation } from "../../../../core/components/chart-layout/hooks/useChartLayoutOrientation.js";
import { getLegendOrientation } from "../../../../core/components/chart-layout/utils/orientation.js";
import { ChartLegend } from "../../../../core/components/legend/categorical/ChartLegend.js";
import { useTooltipReducer } from "../../../../core/components/overlay-tooltip/useTooltipReducer.js";
import { buildCategoricalItemsFromScale } from "../../../../core/utils/build-categorical-items-from-scale.js";
import { safeGet } from "../../../../core/utils/safe-get.js";
import { useXYChartColorScales } from "../../../hooks/useXYChartColorScales.js";
import { useXYChartTruncationMode } from "../../../hooks/useXYChartTruncationMode.js";
import { reduceFilterSeries } from "../../../store/reducer-filter-series.js";
import { useSetState } from "../../../store/selectors.js";
import {
  AREA_VARIANT,
  BAR_VARIANT,
  DOT_VARIANT,
  LINE_VARIANT,
  RECT_VARIANT
} from "../../../types/xy-chart-internals.js";
import { isRangeLegend } from "../../../utils/is-range-legend.js";
import { isRectSeries } from "../../../utils/type-guards.js";
const CHART_ICONS = {
  [LINE_VARIANT]: LINE_CHART_ICON_SHAPE,
  [BAR_VARIANT]: BAR_CHART_ICON_SHAPE,
  [RECT_VARIANT]: RECT_CHART_ICON_SHAPE,
  [AREA_VARIANT]: AREA_CHART_ICON_SHAPE
};
function getLegendItems(series, metadata, isMixedSeries, scaleColorMap, isCategoricalRectOnly) {
  return series.flatMap((series2) => {
    if (series2.type === RECT_VARIANT) {
      if (isRangeLegend(metadata, series2)) {
        return [];
      }
      const scale = safeGet(scaleColorMap, series2.id);
      const overrideItems = safeGet(metadata, series2).categoricalColorOverrides;
      const shape = isCategoricalRectOnly ? "circle" : CHART_ICONS[series2.type];
      const items = buildCategoricalItemsFromScale(
        scale,
        overrideItems,
        //We need to compare exactly the same id when an action is selected
        //In RectSeries, categories are unique per series.
        (category) => `${series2.id}--${category}`
      );
      return items.map((item) => ({
        ...item,
        icon: /* @__PURE__ */ jsx(XYLegendIcon, { color: item.color.Default, shape })
      }));
    } else {
      const { name, color, shape = "circle" } = safeGet(metadata, series2);
      const iconShape = isMixedSeries && series2.type !== DOT_VARIANT ? CHART_ICONS[series2.type] : shape;
      return {
        id: series2.id,
        color,
        name,
        icon: /* @__PURE__ */ jsx(XYLegendIcon, { color: color.Default, shape: iconShape })
      };
    }
  });
}
function XYLegend({ series, seriesToMetadata }) {
  const dispatch = useTooltipReducer();
  const { orientation: layoutOrientation } = useChartLayoutOrientation();
  const orientation = getLegendOrientation(layoutOrientation);
  const setState = useSetState();
  const isLoading = useChartLayoutLoading();
  const scaleColorMap = useXYChartColorScales();
  const { truncationMode } = useXYChartTruncationMode();
  const firstSeries = series[0];
  const isCategoricalRectOnly = isRectSeries(firstSeries) && series.every(isRectSeries) && !isRangeLegend(seriesToMetadata, firstSeries);
  const isMixedSeries = useMemo(() => {
    const typeSet = new Set(series.map((s) => s.type));
    return typeSet.size >= 2;
  }, [series]);
  const legendItems = useMemo(
    () => getLegendItems(
      series,
      seriesToMetadata,
      isMixedSeries,
      scaleColorMap,
      isCategoricalRectOnly
    ),
    // eslint-disable-next-line react-hooks/exhaustive-deps, react-hooks/preserve-manual-memoization
    [series, seriesToMetadata]
  );
  const legendActions = getXYLegendActions(series, seriesToMetadata);
  const updateVisibleSeries = useCallback(
    (visibleItems) => {
      const visibleIds = new Set(visibleItems.map((item) => item.id));
      dispatch({ type: "RESET_TOOLTIP" })();
      setState(reduceFilterSeries(visibleIds));
    },
    [setState, dispatch]
  );
  const updateHighlightedSeriesId = (legendItemId) => {
    setState((prev) => ({ ...prev, highlightedSeriesId: legendItemId }));
  };
  return /* @__PURE__ */ jsx(
    ChartLegend,
    {
      items: legendItems,
      orientation,
      truncationMode,
      interactionsDisabled: isLoading,
      onSeriesAction: legendActions,
      onHover: (legendItem) => updateHighlightedSeriesId(legendItem?.id),
      onClick: updateVisibleSeries
    }
  );
}
export {
  XYLegend
};
