// @dynatrace/strato-components/esm/overlays/sheet/Sheet.js
import { jsx, jsxs } from "react/jsx-runtime";
import { AnimatePresence, motion } from "framer-motion";
import {
  forwardRef,
  useMemo,
  useRef
} from "react";
import AnimationsJS from "@dynatrace/strato-design-tokens/animations-js";
import { calculateBreakpointFromContainerSize } from "./sheet-container-utils.js";
import {
  sheetBackdropCSS,
  sheetCSS,
  sheetHeadingCSS,
  sheetScrollableContainerCSS,
  sheetWrapperCSS
} from "./Sheet.sty.js";
import { FocusScope } from "../../core/components/focus-scope/FocusScope.js";
import { OverlayContainer } from "../../core/components/OverlayContainer.js";
import { useAriaLabelingProps as _useAriaLabelingProps } from "../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useContainerSize } from "../../core/hooks/useContainerSize.js";
import { useId } from "../../core/hooks/useId.js";
import { useMergeRefs } from "../../core/hooks/useMergeRefs.js";
import { useOverlayStacking } from "../../core/hooks/useOverlayStacking.js";
import { OverlayStackingProvider } from "../../core/providers/OverlayStackingProvider.js";
import { attemptFocus as _attemptFocus } from "../../core/utils/focus-management/attempt-focus.js";
import { getFirstFocusableChild as _getFirstFocusableChild } from "../../core/utils/focus-management/get-first-focusable-child.js";
import {
  fadeInEasing,
  fadeOutEasing,
  slideInEasing,
  slideOutEasing
} from "../../core/utils/framer-motion-easings.js";
import { mergeProps as _mergeProps } from "../../core/utils/merge-props.js";
import { Flex } from "../../layouts/flex/Flex.js";
import { Heading } from "../../typography/heading/Heading.js";
import { useDialog } from "../hooks/useDialog.js";
import { useOverlay } from "../hooks/useOverlay.js";
const COMPONENT_NAME = "Sheet";
const getAnimationVariants = (topOffset) => ({
  backdrop: {
    enterBackdrop: {
      // For the opacity we need to use the default value here. Framer motion would complain
      // about too complex values, when transitioning from two custom property values within
      // an rgba value apparently.
      opacity: 1,
      transition: {
        duration: AnimationsJS.FadeIn.Duration.Fast,
        ease: fadeInEasing
      }
    },
    exitBackdrop: {
      opacity: 0,
      transition: {
        duration: AnimationsJS.FadeOut.Duration.Fast,
        ease: fadeOutEasing,
        delay: AnimationsJS.FadeOut.Duration.Base
      }
    }
  },
  sheet: {
    enterSlide: {
      top: topOffset,
      opacity: 1,
      transition: {
        duration: AnimationsJS.SlideIn.Duration.Slow,
        ease: slideInEasing
      }
    },
    exitSlide: () => ({
      top: `calc(${topOffset}px + 10%)`,
      opacity: 0.3,
      transition: {
        duration: AnimationsJS.SlideOut.Duration.Slow,
        ease: slideOutEasing
      }
    })
  },
  content: {
    enterFade: {
      opacity: 1,
      transition: {
        duration: AnimationsJS.SlideIn.Duration.Slow,
        ease: fadeInEasing
      }
    },
    exitFade: {
      opacity: 0,
      transition: {
        duration: AnimationsJS.SlideOut.Duration.Slow,
        ease: fadeOutEasing
      }
    }
  }
});
const Sheet = /* @__PURE__ */ forwardRef(
  (props, forwardedRef) => {
    const {
      show = false,
      title,
      children,
      onDismiss,
      actions,
      topoffset = 28,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...remainingProps
    } = props;
    const { "aria-labelledby": ariaLabelledBy, ...ariaLabelingProps } = _useAriaLabelingProps(remainingProps);
    const ref = useRef(null);
    const animationVariants = getAnimationVariants(topoffset);
    const { dialogProps } = useDialog(remainingProps);
    const { overlayProps } = useOverlay(
      {
        ...props,
        onClose: () => {
          onDismiss?.();
        },
        isOpen: show,
        isDismissible: false,
        isKeyboardDismissDisabled: false
      },
      ref
    );
    const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
    const childrenScopeRef = useRef(null);
    const actionScopeRef = useRef(null);
    const focusFirstChildElement = () => {
      if (childrenScopeRef.current && actionScopeRef.current) {
        let firstFocusableChild = _getFirstFocusableChild(
          childrenScopeRef.current
        );
        if (!firstFocusableChild) {
          firstFocusableChild = _getFirstFocusableChild(actionScopeRef.current);
        }
        firstFocusableChild && _attemptFocus(firstFocusableChild);
      }
    };
    const containerRef = useRef(null);
    const containerSize = useContainerSize(containerRef);
    const mergedRef = useMergeRefs([containerRef, forwardedRef]);
    const breakpointStyle = useMemo(
      () => calculateBreakpointFromContainerSize(containerSize),
      [containerSize]
    );
    const overlayContainerProps = useOverlayStacking(0, "overlay");
    const headingId = useId();
    return /* @__PURE__ */ jsx(
      OverlayContainer,
      {
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        style: { ...overlayContainerProps.style, ...consumerStyle },
        className: consumerClassName,
        ...restProps,
        ...behavioralTrackingProps,
        "data-dt-component": COMPONENT_NAME,
        children: /* @__PURE__ */ jsx(AnimatePresence, { children: show && /* @__PURE__ */ jsx(
          motion.div,
          {
            variants: animationVariants.backdrop,
            initial: "exitBackdrop",
            animate: "enterBackdrop",
            exit: "exitBackdrop",
            className: sheetWrapperCSS,
            children: /* @__PURE__ */ jsx(
              "div",
              {
                ref: mergedRef,
                onPointerDown: (event) => {
                  if (event.target === event.currentTarget) {
                    onDismiss?.();
                  }
                },
                className: sheetBackdropCSS,
                children: /* @__PURE__ */ jsx(
                  FocusScope,
                  {
                    contain: true,
                    children: /* @__PURE__ */ jsx(
                      motion.div,
                      {
                        ..._mergeProps(overlayProps, dialogProps, {
                          style: breakpointStyle
                          // eslint-disable-next-line @typescript-eslint/no-explicit-any
                        }),
                        className: sheetCSS,
                        "data-ismodal": true,
                        variants: animationVariants.sheet,
                        initial: "exitSlide",
                        animate: "enterSlide",
                        exit: "exitSlide",
                        onAnimationComplete: () => {
                          show && focusFirstChildElement();
                          window.dispatchEvent(new Event("animationDone"));
                        },
                        ref,
                        "aria-labelledby": ariaLabelledBy ?? headingId,
                        ...ariaLabelingProps,
                        children: /* @__PURE__ */ jsx(
                          motion.div,
                          {
                            className: sheetScrollableContainerCSS,
                            style: { height: `calc(100% - ${topoffset}px)` },
                            variants: animationVariants.content,
                            initial: "exitFade",
                            animate: "enterFade",
                            exit: "exitFade",
                            children: /* @__PURE__ */ jsx(OverlayStackingProvider, { children: /* @__PURE__ */ jsxs(Flex, { gap: 16, flexDirection: "column", height: "100%", children: [
                              (title || actions) && /* @__PURE__ */ jsxs(
                                Flex,
                                {
                                  justifyContent: title ? "space-between" : "flex-end",
                                  alignItems: "flex-start",
                                  children: [
                                    title && /* @__PURE__ */ jsx(
                                      Heading,
                                      {
                                        level: 5,
                                        as: "h2",
                                        id: headingId,
                                        className: sheetHeadingCSS,
                                        children: title
                                      }
                                    ),
                                    actions && /* @__PURE__ */ jsx(
                                      Flex,
                                      {
                                        flexWrap: "wrap",
                                        justifyContent: "flex-end",
                                        ref: actionScopeRef,
                                        children: actions
                                      }
                                    )
                                  ]
                                }
                              ),
                              /* @__PURE__ */ jsx(
                                "div",
                                {
                                  ref: childrenScopeRef,
                                  style: {
                                    position: "relative",
                                    height: "100%",
                                    minHeight: 0
                                  },
                                  children
                                }
                              )
                            ] }) })
                          }
                        )
                      }
                    )
                  }
                )
              }
            )
          }
        ) })
      }
    );
  }
);
Sheet.displayName = COMPONENT_NAME;
export {
  Sheet
};
