// @dynatrace/strato-components/charts/timeseries/components/thresholds/ThresholdIndicators.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ThresholdIndicators_exports = {};
__export(ThresholdIndicators_exports, {
  ThresholdIndicators: () => ThresholdIndicators
});
module.exports = __toCommonJS(ThresholdIndicators_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_ThresholdIndicators_css = require("./ThresholdIndicators.sty.js");
var import_useGridPlotSize = require("../../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_ThresholdIndicator = require("../../../core/components/threshold-bar/indicators/ThresholdIndicator.js");
var import_type_guards = require("../../../core/types/type-guards.js");
var import_thresholds_to_indicator = require("../../../core/utils/thresholds-to-indicator.js");
var import_selectors = require("../../hooks/store/selectors.js");
var import_useThresholdConfig = require("../../hooks/useThresholdConfig.js");
var import_useXScale = require("../../hooks/useXScale.js");
var import_useYScales = require("../../hooks/useYScales.js");
function getThresholdIndicatorType(data, strokeOnly) {
  const isRange = (0, import_type_guards.isStaticRange)(data) || (0, import_type_guards.isDynamicRange)(data);
  if (isRange && !strokeOnly) {
    return "range";
  }
  return "line";
}
const ThresholdIndicators = ({
  indicatorType
}) => {
  const { left, right } = (0, import_useYScales.useYScales)();
  const { currentScale: xScale } = (0, import_useXScale.useXScales)();
  const { left: leftThresholds, right: rightThresholds } = (0, import_useThresholdConfig.useThresholdConfig)();
  const { width } = (0, import_useGridPlotSize.useGridPlotSize)();
  const { hoveredThresholdId } = (0, import_selectors.useThresholdsState)();
  const thresholds = [
    ...leftThresholds ?? [],
    ...rightThresholds ?? []
  ];
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, { children: thresholds.map(({ color, data, strokeOnly, position, id }) => {
    if (getThresholdIndicatorType(data, strokeOnly) !== indicatorType) {
      return null;
    }
    const yScale = position === "left" ? left : right;
    if (!yScale) {
      return null;
    }
    const indicatorData = (0, import_thresholds_to_indicator.transformThresholdIntoIndicator)(data, width, yScale.scale, xScale);
    const isStrokeOnly = (0, import_type_guards.isStatic)(data) || (0, import_type_guards.isDynamic)(data) || strokeOnly;
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "svg",
      {
        className: indicatorType === "range" ? import_ThresholdIndicators_css.thresholdPlotRange : import_ThresholdIndicators_css.thresholdPlotLine,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_ThresholdIndicator.ThresholdIndicator,
          {
            data: indicatorData,
            color,
            strokeOnly: isStrokeOnly,
            isActive: hoveredThresholdId === id
          }
        )
      },
      `threshold-indicator-${id}`
    );
  }) });
};
