// @dynatrace/strato-components/charts/timeseries/components/thresholds/TimeseriesThresholdBar.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TimeseriesThresholdBar_exports = {};
__export(TimeseriesThresholdBar_exports, {
  TimeseriesThresholdBar: () => TimeseriesThresholdBar
});
module.exports = __toCommonJS(TimeseriesThresholdBar_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_ThresholdBar = require("../../../core/components/threshold-bar/ThresholdBar.js");
var import_threshold_to_marker = require("../../../core/utils/threshold-to-marker.js");
var import_selectors = require("../../hooks/store/selectors.js");
var import_useThresholdConfig = require("../../hooks/useThresholdConfig.js");
var import_useTimeseriesYFormatter = require("../../hooks/useTimeseriesYFormatter.js");
var import_useXScale = require("../../hooks/useXScale.js");
var import_useYScales = require("../../hooks/useYScales.js");
function TimeseriesThresholdBar(props) {
  const { position, length } = props;
  const { left: leftThreshold, right: rightThreshold } = (0, import_useThresholdConfig.useThresholdConfig)();
  const { left, right } = (0, import_useYScales.useYScales)();
  const { currentScale } = (0, import_useXScale.useXScales)();
  const [leftFormatter, rightFormatter] = (0, import_useTimeseriesYFormatter.useTimeseriesYFormatter)();
  const setState = (0, import_selectors.useSetState)();
  const isFlipped = position !== "left";
  const yScale = position === "left" ? left : right;
  if (!yScale) {
    return null;
  }
  const thresholds = position === "left" ? leftThreshold : rightThreshold;
  if (!thresholds || !thresholds.length) {
    return null;
  }
  const axisFormatter = position === "left" ? leftFormatter : rightFormatter;
  const markers = thresholds.reduce(
    (0, import_threshold_to_marker.transformThresholdsToMarkers)(
      position,
      currentScale,
      yScale.scale,
      axisFormatter
    ),
    []
  );
  const handleHover = (id) => {
    setState((prev) => ({
      ...prev,
      thresholds: {
        ...thresholds,
        hoveredThresholdId: id
      }
    }));
  };
  return markers.length ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ThresholdBar.ThresholdBar,
    {
      markers,
      length,
      layout: "vertical",
      isFlipped,
      onHover: handleHover
    }
  ) : null;
}
TimeseriesThresholdBar["displayName"] = "TimeseriesThresholdBar";
