// @dynatrace/strato-components/charts/xy-chart/utils/axis-interactions.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var axis_interactions_exports = {};
__export(axis_interactions_exports, {
  interactionsFor: () => interactionsFor,
  isInfiniteZoomEnabledFor: () => isInfiniteZoomEnabledFor,
  isPanEnabledFor: () => isPanEnabledFor,
  isZoomEnabledFor: () => isZoomEnabledFor
});
module.exports = __toCommonJS(axis_interactions_exports);
var import_axis = require("../types/axis.js");
var import_guards = require("../types/guards.js");
const interactionsFor = (axis) => {
  if ((0, import_guards.isCategorical)(import_axis.XYAxisMapper[axis.type])) {
    return {
      zoom: false,
      pan: false,
      infinite: false
    };
  }
  return {
    zoom: !axis.disableZoom,
    pan: !axis.disablePan,
    infinite: !!axis.infiniteZoom,
    initial: axis.initialZoom,
    current: axis.currentZoom
  };
};
const isZoomEnabledFor = (axis) => {
  return axis.interactions.zoom;
};
const isPanEnabledFor = (axis) => {
  return axis.interactions.pan;
};
const isInfiniteZoomEnabledFor = (axis) => {
  return axis.interactions.infinite;
};
