// @dynatrace/strato-components/charts/xy-chart/utils/build-series-action-item.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var build_series_action_item_exports = {};
__export(build_series_action_item_exports, {
  buildAreaSeriesActionItem: () => buildAreaSeriesActionItem,
  buildBarSeriesActionItem: () => buildBarSeriesActionItem
});
module.exports = __toCommonJS(build_series_action_item_exports);
var import_safe_get = require("../../core/utils/safe-get.js");
var import_guards = require("../types/guards.js");
var import_xy_chart = require("../types/xy-chart.js");
function baselineDirectionToAccessor(baseline) {
  return baseline ? `${baseline}0` : void 0;
}
function getBaselineCategoryInfo(datapoint, baselineAccessor, metadata) {
  const { xAxis, yAxis } = metadata;
  const axis = baselineAccessor === "x0" ? xAxis : yAxis;
  const value = datapoint[baselineAccessor];
  if (!(0, import_guards.isCategoricalAxis)(axis) || typeof value !== "string") {
    return {
      category: void 0,
      isBaselineCategorical: false
    };
  }
  return {
    category: value,
    isBaselineCategorical: true
  };
}
const barSeriesFilterStrategy = (_series, currentMetadata, _seriesToMetadata, haveSameGrowingAxis) => {
  return currentMetadata.variant === import_xy_chart.BAR_VARIANT_STACK && !haveSameGrowingAxis;
};
const areaSeriesFilterStrategy = (series, _currentMetadata, seriesToMetadata, haveSameGrowingAxis) => {
  if (!haveSameGrowingAxis) {
    return true;
  }
  const seriesMetadata = (0, import_safe_get.safeGet)(seriesToMetadata, series);
  return seriesMetadata.variant === "band";
};
const areaSimpleActionItemStrategy = (metadata) => metadata.variant === "band";
const barSimpleActionItemStrategy = (metadata) => metadata.variant === import_xy_chart.BAR_VARIANT_SINGLE;
function buildSeriesActionItemInternal({
  datapoint,
  allSeries,
  seriesToMetadata,
  datapointMetadataMap,
  currentSeries,
  simpleActionItemStrategy,
  seriesFilterStrategy
}) {
  const currentMetadata = (0, import_safe_get.safeGet)(seriesToMetadata, currentSeries);
  const { baseline } = (0, import_safe_get.safeGet)(datapointMetadataMap, datapoint);
  const baselineAccessor = baselineDirectionToAccessor(baseline);
  if (!baselineAccessor || simpleActionItemStrategy(currentMetadata)) {
    return { datapoint, name: currentMetadata.name };
  }
  const { category, isBaselineCategorical } = getBaselineCategoryInfo(
    datapoint,
    baselineAccessor,
    currentMetadata
  );
  if (!isBaselineCategorical || category === void 0) {
    return { datapoint, name: currentMetadata.name };
  }
  const relatedDatapoints = [];
  const primaryAxisAccessor = baseline === "x" ? "xAxisId" : "yAxisId";
  const growingAxisAccessor = baseline === "x" ? "yAxisId" : "xAxisId";
  for (const series of allSeries) {
    const haveSamePrimaryAxis = series[primaryAxisAccessor] === currentSeries[primaryAxisAccessor];
    const haveSameGrowingAxis = series[growingAxisAccessor] === currentSeries[growingAxisAccessor];
    if (!haveSamePrimaryAxis || seriesFilterStrategy(
      series,
      currentMetadata,
      seriesToMetadata,
      haveSameGrowingAxis
    )) {
      continue;
    }
    for (const dp of series.data) {
      if (dp[baselineAccessor] === category) {
        relatedDatapoints.push(dp);
      }
    }
  }
  return {
    datapoint,
    category,
    datapoints: relatedDatapoints,
    name: currentMetadata.name
  };
}
function buildBarSeriesActionItem(params) {
  return buildSeriesActionItemInternal({
    ...params,
    simpleActionItemStrategy: barSimpleActionItemStrategy,
    seriesFilterStrategy: barSeriesFilterStrategy
  });
}
function buildAreaSeriesActionItem(params) {
  return buildSeriesActionItemInternal({
    ...params,
    simpleActionItemStrategy: areaSimpleActionItemStrategy,
    seriesFilterStrategy: areaSeriesFilterStrategy
  });
}
