// @dynatrace/strato-components/charts/xy-chart/utils/calculate-axes-domains.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var calculate_axes_domains_exports = {};
__export(calculate_axes_domains_exports, {
  calculateAxesDomains: () => calculateAxesDomains
});
module.exports = __toCommonJS(calculate_axes_domains_exports);
var import_axis = require("../types/axis.js");
function getMinMax(axisId, coordinate0, coordinate1, axisIdToScaleDomain) {
  const minMax = axisIdToScaleDomain.get(axisId) ?? {
    min: coordinate0,
    max: coordinate1
  };
  const minValues = [minMax.min, coordinate0, coordinate1].filter(
    (value) => value !== void 0
  );
  const maxValues = [minMax.max, coordinate0, coordinate1].filter(
    (value) => value !== void 0
  );
  minMax.min = Math.min(...minValues);
  minMax.max = Math.max(...maxValues);
  return minMax;
}
function getCategoricalDomain(axisId, v0, v1, axisIdToScaleDomain) {
  const categoricalSet = axisIdToScaleDomain.get(axisId) ?? /* @__PURE__ */ new Set();
  if (v0) {
    categoricalSet.add(v0);
  }
  if (v1) {
    categoricalSet.add(v1);
  }
  return categoricalSet;
}
const calculateAxesDomainsMap = {
  [import_axis.XYAxisTypeLinear]: (args) => {
    return getMinMax(
      args.axisId,
      args.a0,
      args.a1,
      args.axisIdToScaleDomain
    );
  },
  [import_axis.XYAxisTypeLog]: (args) => {
    return getMinMax(
      args.axisId,
      args.a0,
      args.a1,
      args.axisIdToScaleDomain
    );
  },
  [import_axis.XYAxisTypeCategorical]: (args) => {
    return getCategoricalDomain(
      args.axisId,
      args.a0,
      args.a1,
      args.axisIdToScaleDomain
    );
  },
  [import_axis.XYAxisTypeTime]: (args) => {
    return getMinMax(
      args.axisId,
      args.a0?.valueOf(),
      args.a1?.valueOf(),
      args.axisIdToScaleDomain
    );
  }
};
function calculateAxesDomains(xAxis, yAxis, datapoint, axisIdToScaleDomain, isSeriesVisible) {
  const xDomain = calculateAxesDomainsMap[xAxis.type]({
    axisId: xAxis.id,
    a0: datapoint.x0,
    a1: datapoint.x1,
    axisIdToScaleDomain
  });
  const yDomain = isSeriesVisible || yAxis.type === import_axis.XYAxisTypeCategorical ? calculateAxesDomainsMap[yAxis.type]({
    axisId: yAxis.id,
    a0: datapoint.y0,
    a1: datapoint.y1,
    axisIdToScaleDomain
  }) : axisIdToScaleDomain.get(yAxis.id) ?? {};
  return { xDomain, yDomain };
}
