// @dynatrace/strato-components/charts/xy-chart/utils/generate-nice-unit-ticks.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var generate_nice_unit_ticks_exports = {};
__export(generate_nice_unit_ticks_exports, {
  computeNiceUnitTicks: () => computeNiceUnitTicks,
  createDoublingStepCandidateBuilder: () => createDoublingStepCandidateBuilder,
  generateTicks: () => generateTicks,
  getBestModulusDivisor: () => getBestModulusDivisor,
  niceUnitDomain: () => niceUnitDomain,
  shouldUseNiceUnitTicks: () => shouldUseNiceUnitTicks
});
module.exports = __toCommonJS(generate_nice_unit_ticks_exports);
var import_units = require("@dynatrace-sdk/units");
var import_formatter = require("../../core/types/formatter.js");
var import_adaptively_reduce_horizontal_ticks = require("../../core/utils/adaptively-reduce-horizontal-ticks.js");
var import_ticks = require("../../core/utils/ticks.js");
var import_guards = require("../types/guards.js");
function shouldUseNiceUnitTicks(config) {
  if (!config || config.isRelative || !(0, import_guards.isLinear)(config.type)) {
    return false;
  }
  const formatter = config.formatter;
  return (0, import_formatter.isFormatterOptions)(formatter) && formatter.input !== void 0 && formatter.output !== void 0;
}
function getBestModulusDivisor(totalDomain) {
  let bestDivisor = 4;
  let minMod = totalDomain % 4;
  for (let divisor = 4; divisor <= 8; divisor++) {
    const mod = totalDomain % divisor;
    if (mod === 0) {
      return divisor;
    }
    if (mod < minMod || mod === minMod && divisor < bestDivisor) {
      minMod = mod;
      bestDivisor = divisor;
    }
  }
  return bestDivisor;
}
function convertDomainToTargetUnit(domain, formatterOptions) {
  const inputUnit = formatterOptions.input;
  const outputUnit = formatterOptions.output;
  return {
    targetMin: (0, import_units.convert)(domain[0], inputUnit, outputUnit, true),
    targetMax: (0, import_units.convert)(domain[1], inputUnit, outputUnit, true),
    inputUnit,
    outputUnit
  };
}
function computeNiceStep(totalDomain, tickDivisor) {
  if (totalDomain < 1) {
    return 0;
  }
  const amount = totalDomain / Math.max(tickDivisor, 1);
  return amount < 10 ? Math.ceil(amount) : Math.floor(amount / 10) * 10;
}
function generateTicks(domain, step) {
  if (step <= 0) {
    return [0];
  }
  const [min, max] = domain;
  const allTicks = [0];
  if (Math.sign(min) === Math.sign(max)) {
    const sign = Math.sign(max) || 1;
    const absMax = Math.max(Math.abs(min), Math.abs(max));
    const count = Math.floor(absMax / step);
    for (let i = 1; i <= count; i++) {
      allTicks.push(step * i * sign);
    }
  } else {
    const posCount = Math.floor(Math.abs(max) / step);
    const negCount = Math.floor(Math.abs(min) / step);
    for (let i = 1; i <= negCount; i++) {
      allTicks.push(-step * i);
    }
    for (let i = 1; i <= posCount; i++) {
      allTicks.push(step * i);
    }
  }
  return allTicks.sort((a, b) => a - b);
}
function niceUnitDomain(domain, formatterOptions, niceMin, niceMax) {
  const { targetMin, targetMax, inputUnit, outputUnit } = convertDomainToTargetUnit(domain, formatterOptions);
  const totalDomain = Math.abs(targetMin) + Math.abs(targetMax);
  const step = computeNiceStep(totalDomain, getBestModulusDivisor(totalDomain));
  if (step === 0) {
    return null;
  }
  const nicedMin = niceMin ? Math.floor(targetMin / step) * step : targetMin;
  const nicedMax = niceMax ? Math.ceil(targetMax / step) * step : targetMax;
  return {
    min: (0, import_units.convert)(nicedMin, outputUnit, inputUnit, true),
    max: (0, import_units.convert)(nicedMax, outputUnit, inputUnit, true)
  };
}
function computeNiceUnitTicks(domain, formatterOptions, availableSize, orientation, formatter, scale) {
  const { targetMin, targetMax, inputUnit, outputUnit } = convertDomainToTargetUnit(domain, formatterOptions);
  if (targetMin === targetMax) {
    return {
      ticks: ["0"],
      formattedTicks: [formatter(0)],
      numericTicks: [0]
    };
  }
  const totalDomain = Math.abs(targetMin) + Math.abs(targetMax);
  let tickDivisor = getBestModulusDivisor(totalDomain);
  const TICK_SPACING = 20;
  if (tickDivisor * TICK_SPACING >= availableSize) {
    tickDivisor = orientation === "vertical" ? (0, import_ticks.getVerticalNumTicks)(availableSize) : (0, import_ticks.getHorizontalNumTicks)(availableSize);
  }
  const step = computeNiceStep(totalDomain, tickDivisor);
  if (step === 0) {
    return null;
  }
  const buildResult = (currentStep) => {
    const targetTicks = generateTicks([targetMin, targetMax], currentStep);
    if (targetTicks[targetTicks.length - 1] !== targetMax) {
      targetTicks.push(targetMax);
    }
    if (targetTicks[0] !== targetMin) {
      targetTicks.unshift(targetMin);
    }
    const baseUnitTicks = targetTicks.map(
      (tick) => (0, import_units.convert)(tick, outputUnit, inputUnit, true)
    );
    return {
      ticks: baseUnitTicks.map(String),
      formattedTicks: baseUnitTicks.map((tick) => formatter(tick)),
      numericTicks: baseUnitTicks
    };
  };
  if (orientation !== "horizontal" || scale === void 0) {
    return buildResult(step);
  }
  const reduced = (0, import_adaptively_reduce_horizontal_ticks.adaptivelyReduceHorizontalTicks)(
    createDoublingStepCandidateBuilder(step, buildResult),
    scale
  );
  return {
    ticks: reduced.ticks,
    formattedTicks: reduced.formattedTicks,
    numericTicks: reduced.numericTicks
  };
}
function createDoublingStepCandidateBuilder(initialStep, buildResult) {
  let currentStep = initialStep;
  return (attempt) => {
    if (attempt > 0) {
      currentStep *= 2;
    }
    const r = buildResult(currentStep);
    return { ...r, rawTicks: r.numericTicks };
  };
}
