// @dynatrace/strato-components/charts/xy-chart/utils/generate-scales-by-axis-id.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var generate_scales_by_axis_id_exports = {};
__export(generate_scales_by_axis_id_exports, {
  generateInitialTimeScale: () => generateInitialTimeScale,
  generateScalesByAxisId: () => generateScalesByAxisId
});
module.exports = __toCommonJS(generate_scales_by_axis_id_exports);
var import_d3_scale = require("d3-scale");
var import_generate_nice_unit_ticks = require("./generate-nice-unit-ticks.js");
var import_is_axis = require("./is-axis.js");
var import_value_representation = require("./value-representation.js");
var import_constants = require("../../core/constants.js");
var import_core_scale = require("../../core/utils/core-scale.js");
var import_get_min_max_for_scale = require("../../core/utils/get-min-max-for-scale.js");
var import_safe_get = require("../../core/utils/safe-get.js");
var import_guards = require("../types/guards.js");
function generateInitialNumericalScale(initialDomain, axis, isXCoordinate, width, height, minMaxDomain) {
  const scaleType = axis.type.split(".")[1];
  let min, max;
  const initialDomainWithRelativeFilter = axis.isRelative ? {
    min: (0, import_value_representation.toRelative)(initialDomain.min, minMaxDomain),
    max: (0, import_value_representation.toRelative)(initialDomain.max, minMaxDomain)
  } : initialDomain;
  if (axis.interactions.initial) {
    ({ min, max } = initialDomainWithRelativeFilter);
  } else {
    const minMaxByConfig = { min: axis.min, max: axis.max };
    const niceMin = (0, import_guards.isLogarithmicAxis)(axis) ? false : axis.min === "auto";
    const niceMax = (0, import_guards.isLogarithmicAxis)(axis) ? false : axis.max === "auto";
    const rawMinMax = (0, import_core_scale.getMinMax)(
      scaleType,
      initialDomainWithRelativeFilter,
      minMaxByConfig,
      { niceMin: false, niceMax: false }
    );
    const axisFormatter = axis.formatter;
    const niceDomain = (0, import_generate_nice_unit_ticks.shouldUseNiceUnitTicks)(axis) && (niceMin || niceMax) ? (0, import_generate_nice_unit_ticks.niceUnitDomain)(
      [rawMinMax.min, rawMinMax.max],
      axisFormatter,
      niceMin,
      niceMax
    ) : null;
    ({ min, max } = niceDomain ?? (0, import_core_scale.getMinMax)(scaleType, initialDomainWithRelativeFilter, minMaxByConfig, {
      niceMin,
      niceMax
    }));
  }
  const domain = axis.reversed ? [max, min] : [min, max];
  const range = isXCoordinate ? [0, width] : [height, 0];
  return (0, import_core_scale.getScale)(scaleType, domain, range);
}
function generateNumericScale(initialScale, zoomedDomain, scaleType) {
  const range = initialScale.range();
  let domain = initialScale.domain();
  if (zoomedDomain) {
    const min = zoomedDomain.min;
    const max = zoomedDomain.max;
    domain = [min, max];
  }
  return (0, import_core_scale.getScale)(scaleType, domain, range);
}
function generateTimeScale(initialScale, zoomedDomain) {
  const range = initialScale.range();
  let domain;
  if (zoomedDomain) {
    const min = zoomedDomain.min;
    const max = zoomedDomain.max;
    domain = [min, max];
  } else {
    domain = initialScale.domain();
  }
  return (0, import_d3_scale.scaleTime)().domain(domain).range(range);
}
function generateInitialTimeScale(initialDomain, axis, isXCoordinate, width, height) {
  let min, max;
  if (axis.interactions.initial) {
    ({ min, max } = initialDomain);
  } else {
    const minMaxByConfig = { min: axis.min, max: axis.max };
    ({ min, max } = (0, import_get_min_max_for_scale.resolveTimeMinMax)(initialDomain, minMaxByConfig));
  }
  const domain = axis.reversed ? [max, min] : [min, max];
  const range = isXCoordinate ? [0, width] : [height, 0];
  const scale = (0, import_d3_scale.scaleTime)().domain(domain).range(range);
  return scale;
}
function generateCategoricalScale(categoricalValues, axis, isXCoordinate, width, height) {
  const range = isXCoordinate ? [0, width] : [0, height];
  const domain = axis.reversed ? Array.from(categoricalValues).reverse() : categoricalValues;
  return (0, import_d3_scale.scaleBand)(domain, range);
}
function generateScalesByAxisId(axes, axisIdToScaleDomain, axisIdToZoomedDomain, axisToDimension, width, height) {
  const axisIdToScales = /* @__PURE__ */ new Map();
  axes.forEach((axis) => {
    const isXCoordinate = (0, import_is_axis.isXAxis)(axis);
    const minMaxDomain = (0, import_safe_get.safeGet)(axisIdToScaleDomain, axis.id);
    if ((0, import_guards.isNumericalAxis)(axis)) {
      const initialDomain = (axis.interactions.initial && {
        min: axis.interactions.initial[0],
        max: axis.interactions.initial[1]
      }) ?? minMaxDomain;
      const initialScale = generateInitialNumericalScale(
        initialDomain,
        axis,
        isXCoordinate,
        width,
        height,
        minMaxDomain
      );
      const zoomedDomain = axisIdToZoomedDomain?.get(axis.id);
      const currentScale = generateNumericScale(
        initialScale,
        zoomedDomain,
        axis.type.split(".")[1]
      );
      axisIdToScales.set(axis.id, { currentScale, initialScale });
    } else if ((0, import_guards.isTimeAxis)(axis)) {
      const initialDomain = (axis.interactions.initial && {
        min: axis.interactions.initial[0],
        max: axis.interactions.initial[1]
      }) ?? (0, import_safe_get.safeGet)(axisIdToScaleDomain, axis.id);
      const initialScale = generateInitialTimeScale(
        initialDomain,
        axis,
        isXCoordinate,
        width,
        height
      );
      const zoomedDomain = axisIdToZoomedDomain?.get(axis.id);
      const currentScale = generateTimeScale(
        initialScale,
        zoomedDomain
      );
      axisIdToScales.set(axis.id, { currentScale, initialScale });
    } else {
      const initialDomain = (0, import_safe_get.safeGet)(
        axisIdToScaleDomain,
        axis.id
      );
      const scale = generateCategoricalScale(
        initialDomain,
        axis,
        isXCoordinate,
        width,
        height
      );
      const dimensions = axisToDimension.get(axis.id);
      axisIdToScales.set(axis.id, {
        initialScale: scale,
        currentScale: scale,
        dimensionScale: dimensions ? (0, import_d3_scale.scaleBand)(dimensions, [
          0,
          scale.copy().paddingInner(import_constants.CATEGORICAL_BARS_INNER_PADDING).paddingOuter(import_constants.CATEGORICAL_BARS_OUTER_PADDING).bandwidth()
        ]).paddingInner(import_constants.CATEGORICAL_BARS_INNER_PADDING) : void 0,
        categoricalBarScale: scale.copy().paddingInner(import_constants.CATEGORICAL_BARS_INNER_PADDING).paddingOuter(import_constants.CATEGORICAL_BARS_OUTER_PADDING)
      });
    }
  });
  return axisIdToScales;
}
