// @dynatrace/strato-components/charts/xy-chart/utils/get-padding-by-axes.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var get_padding_by_axes_exports = {};
__export(get_padding_by_axes_exports, {
  getPaddingByAxes: () => getPaddingByAxes
});
module.exports = __toCommonJS(get_padding_by_axes_exports);
var import_constants = require("../../core/components/data-point/constants.js");
const AXES = ["top", "bottom", "left", "right"];
const Paddings = () => ({
  top: 0,
  bottom: 0,
  left: 0,
  right: 0,
  horizontal: function() {
    return this.left + this.right;
  },
  vertical: function() {
    return this.top + this.bottom;
  }
});
const getPaddingByAxes = (axes, padding = import_constants.OVERLAY_PADDING) => {
  return AXES.reduce((acc, axis) => {
    acc[axis] = axes[axis] && !axes[axis].hidden ? 0 : padding;
    return acc;
  }, Paddings());
};
