// @dynatrace/strato-components/charts/xy-chart/utils/get-ticks-formatted-by-config.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var get_ticks_formatted_by_config_exports = {};
__export(get_ticks_formatted_by_config_exports, {
  createHalvingHintCandidateBuilder: () => createHalvingHintCandidateBuilder,
  getTicksFormattedByConfig: () => getTicksFormattedByConfig
});
module.exports = __toCommonJS(get_ticks_formatted_by_config_exports);
var import_adaptively_reduce_horizontal_ticks = require("../../core/utils/adaptively-reduce-horizontal-ticks.js");
var import_formatter_by_ticks = require("../../core/utils/formatter-by-ticks.js");
var import_guards = require("../types/guards.js");
const MIN_TICK_COUNT = 2;
function getTicksFormattedByConfig(config, numTicks, scale, formatter, availableWidth) {
  const axisType = config?.type || "";
  if ((0, import_guards.isCategorical)(axisType)) {
    return (0, import_formatter_by_ticks.getCategoricalTicks)(scale, formatter);
  }
  const getTicksForCount = (0, import_guards.isLogarithmic)(axisType) ? (count) => (0, import_formatter_by_ticks.getFormattedTicks)(
    count,
    scale,
    void 0,
    formatter
  ) : (count) => (0, import_formatter_by_ticks.getDateNumericalFormattedTicks)(
    count,
    scale,
    (0, import_guards.isLinear)(axisType) && !config?.isRelative ? config?.unit : void 0,
    formatter
  );
  const result = getTicksForCount(numTicks);
  if (availableWidth === void 0 || scale === void 0) {
    return result;
  }
  let initialResult = result;
  if (result.formattedTicks.length < MIN_TICK_COUNT) {
    const boosted = getTicksForCount(MIN_TICK_COUNT);
    if (boosted.formattedTicks.length >= MIN_TICK_COUNT) {
      initialResult = boosted;
    } else {
      return result;
    }
  }
  const continuousScale = scale;
  const rawDomain = typeof scale.domain === "function" ? scale.domain() : null;
  const domainBounds = rawDomain !== null && rawDomain.length >= 2 ? [+rawDomain[0], +rawDomain[rawDomain.length - 1]] : void 0;
  const reduced = (0, import_adaptively_reduce_horizontal_ticks.adaptivelyReduceHorizontalTicks)(
    (0, import_guards.isLogarithmic)(axisType) ? createLogMajorTickThinningBuilder(initialResult, MIN_TICK_COUNT) : createHalvingHintCandidateBuilder(
      initialResult,
      getTicksForCount,
      MIN_TICK_COUNT
    ),
    continuousScale,
    { domainBounds }
  );
  return {
    ...reduced,
    numericTicks: reduced.rawTicks.map(Number)
  };
}
function createLogMajorTickThinningBuilder(initialResult, minTickCount) {
  const allTicks = initialResult.ticks;
  const majorIndices = initialResult.formattedTicks.reduce(
    (acc, label, i) => {
      if (label !== "") {
        acc.push(i);
      }
      return acc;
    },
    []
  );
  return (attempt) => {
    if (attempt === 0) {
      return {
        ...initialResult,
        rawTicks: allTicks
      };
    }
    const step = Math.pow(2, attempt);
    const selectedMajorIndices = majorIndices.filter(
      (_, i) => i === 0 || i === majorIndices.length - 1 || i % step === 0
    );
    if (selectedMajorIndices.length < minTickCount) {
      return null;
    }
    const selectedSet = new Set(selectedMajorIndices);
    const newFormattedTicks = initialResult.formattedTicks.map((label, i) => {
      if (label === "") {
        return "";
      }
      return selectedSet.has(i) ? label : "";
    });
    return {
      ...initialResult,
      formattedTicks: newFormattedTicks,
      rawTicks: allTicks
    };
  };
}
function createHalvingHintCandidateBuilder(initialResult, getTicksForCount, minTickCount) {
  let hint = initialResult.formattedTicks.length;
  return (attempt) => {
    if (attempt === 0) {
      return {
        ...initialResult,
        rawTicks: initialResult.ticks
      };
    }
    hint = Math.max(Math.floor(hint / 2), minTickCount);
    const next = getTicksForCount(hint);
    return { ...next, rawTicks: next.ticks };
  };
}
