// @dynatrace/strato-components/charts/xy-chart/utils/get-visible-axes.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var get_visible_axes_exports = {};
__export(get_visible_axes_exports, {
  getVisibleAxes: () => getVisibleAxes
});
module.exports = __toCommonJS(get_visible_axes_exports);
var import_is_axis = require("./is-axis.js");
function shouldDisplayBothVertical(axes, topAxis, bottomAxis, bothAxis) {
  if (!bothAxis) {
    return false;
  }
  const topIndex = axes.findIndex((axis) => (0, import_is_axis.isTopAxis)(axis) && !axis.hidden);
  const bottomIndex = axes.findIndex(
    (axis) => (0, import_is_axis.isBottomAxis)(axis) && !axis.hidden
  );
  const bothIndex = axes.findIndex((axis) => (0, import_is_axis.isBothAxis)(axis) && !axis.hidden);
  const nonDefined = !topAxis && !bottomAxis;
  return nonDefined && bothAxis || (bothIndex < topIndex || topIndex === -1) && (bothIndex < bottomIndex || bottomIndex === -1);
}
function shouldDisplayBothHorizontal(axes, leftAxis, rightAxis, bothAxis) {
  if (!bothAxis) {
    return false;
  }
  const rightIndex = axes.findIndex(
    (axis) => (0, import_is_axis.isRightAxis)(axis) && !axis.hidden
  );
  const leftIndex = axes.findIndex((axis) => (0, import_is_axis.isLeftAxis)(axis) && !axis.hidden);
  const bothIndex = axes.findIndex((axis) => (0, import_is_axis.isBothAxis)(axis) && !axis.hidden);
  const nonDefined = !leftAxis && !rightAxis;
  return nonDefined && bothAxis || (bothIndex < leftIndex || leftIndex === -1) && (bothIndex < rightIndex || rightIndex === -1);
}
function getVisibleAxes(axes) {
  const topAxis = axes.find((axis) => (0, import_is_axis.isTopAxis)(axis) && !axis.hidden);
  const bottomAxis = axes.find((axis) => (0, import_is_axis.isBottomAxis)(axis) && !axis.hidden);
  const leftAxis = axes.find((axis) => (0, import_is_axis.isLeftAxis)(axis) && !axis.hidden);
  const rightAxis = axes.find((axis) => (0, import_is_axis.isRightAxis)(axis) && !axis.hidden);
  const verticalBothAxis = axes.find(
    (axis) => (0, import_is_axis.isXAxis)(axis) && (0, import_is_axis.isBothAxis)(axis) && !axis.hidden
  );
  const horizontalBothAxis = axes.find(
    (axis) => (0, import_is_axis.isYAxis)(axis) && (0, import_is_axis.isBothAxis)(axis) && !axis.hidden
  );
  const displayBothVertical = shouldDisplayBothVertical(
    axes,
    topAxis,
    bottomAxis,
    verticalBothAxis
  );
  const displayBothHorizontal = shouldDisplayBothHorizontal(
    axes,
    leftAxis,
    rightAxis,
    horizontalBothAxis
  );
  const visibleAxes = {
    top: displayBothVertical ? verticalBothAxis : topAxis,
    bottom: displayBothVertical ? verticalBothAxis : bottomAxis,
    left: displayBothHorizontal ? horizontalBothAxis : leftAxis,
    right: displayBothHorizontal ? horizontalBothAxis : rightAxis
  };
  return visibleAxes;
}
