// @dynatrace/strato-components/charts/xy-chart/utils/grid-templates.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var grid_templates_exports = {};
__export(grid_templates_exports, {
  AREAS: () => AREAS,
  gridTemplateAreas: () => gridTemplateAreas,
  gridTemplateColumns: () => gridTemplateColumns,
  gridTemplateRows: () => gridTemplateRows
});
module.exports = __toCommonJS(grid_templates_exports);
var import_constants = require("../../core/components/axes/constants.js");
const gridTemplateColumns = (leftAxisMaxWidth = "auto", rightAxisMaxWidth = "auto") => {
  const leftMax = typeof leftAxisMaxWidth === "number" ? `${Math.max(leftAxisMaxWidth, import_constants.MIN_AXIS_TRUNCATION_SIZE)}px` : leftAxisMaxWidth;
  const rightMax = typeof rightAxisMaxWidth === "number" ? `${Math.max(rightAxisMaxWidth, import_constants.MIN_AXIS_TRUNCATION_SIZE)}px` : rightAxisMaxWidth;
  return `minmax(0, ${leftMax}) auto minmax(60%, 1fr) auto minmax(0, ${rightMax})`;
};
const gridTemplateRows = (bottomAxisMaxHeight = "auto", topAxisMaxHeight = "auto", annotationsArea = "auto") => {
  const maxTop = typeof topAxisMaxHeight === "number" ? `${Math.max(topAxisMaxHeight, import_constants.MIN_AXIS_TRUNCATION_SIZE)}px` : topAxisMaxHeight;
  const maxBottom = typeof bottomAxisMaxHeight === "number" ? `${Math.max(bottomAxisMaxHeight, import_constants.MIN_AXIS_TRUNCATION_SIZE)}px` : bottomAxisMaxHeight;
  const annotationsAreaSize = typeof annotationsArea === "number" ? `${annotationsArea}px` : "auto";
  return `minmax(0, ${annotationsAreaSize}) minmax(0, ${maxTop}) minmax(60%, 1fr) minmax(0, ${maxBottom})`;
};
const AREAS = {
  annotations: "annotations",
  topAxis: "top-axis",
  bottomAxis: "bottom-axis",
  leftAxis: "left-axis",
  rightAxis: "right-axis",
  leftThreshold: "left-threshold",
  rightThreshold: "right-threshold"
};
const gridTemplateAreas = `
  "${AREAS.annotations} ${AREAS.annotations} ${AREAS.annotations} . ."
  ". . ${AREAS.topAxis} . ."
  "${AREAS.leftAxis} ${AREAS.leftThreshold} plot ${AREAS.rightThreshold} ${AREAS.rightAxis}"
  ". . ${AREAS.bottomAxis} . ."
`;
