// @dynatrace/strato-components/charts/xy-chart/utils/is-axis.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var is_axis_exports = {};
__export(is_axis_exports, {
  isBothAxis: () => isBothAxis,
  isBottomAxis: () => isBottomAxis,
  isLeftAxis: () => isLeftAxis,
  isRightAxis: () => isRightAxis,
  isTopAxis: () => isTopAxis,
  isXAxis: () => isXAxis,
  isYAxis: () => isYAxis
});
module.exports = __toCommonJS(is_axis_exports);
var import_xy_chart = require("../types/xy-chart.js");
function isXAxis(axis) {
  return axis.coordinate === "x";
}
function isYAxis(axis) {
  return axis.coordinate === "y";
}
function isBottomAxis(axis) {
  return axis.position === import_xy_chart.BOTTOM_POSITION;
}
function isTopAxis(axis) {
  return axis.position === import_xy_chart.TOP_POSITION;
}
function isLeftAxis(axis) {
  return axis.position === import_xy_chart.LEFT_POSITION;
}
function isRightAxis(axis) {
  return axis.position === import_xy_chart.RIGHT_POSITION;
}
function isBothAxis(axis) {
  return axis.position === import_xy_chart.BOTH_POSITION;
}
