// @dynatrace/strato-components/charts/xy-chart/utils/iterate-slots.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var iterate_slots_exports = {};
__export(iterate_slots_exports, {
  iterateSlots: () => iterateSlots
});
module.exports = __toCommonJS(iterate_slots_exports);
var import_react = require("react");
var import_iterate_xy_chart_annotations = require("./iterate-xy-chart-annotations.js");
var import_Intent = require("../../../core/slots/Intent/Intent.js");
var import_is_slot = require("../../../core/utils/is-slot.js");
var import_ColorRule = require("../../core/slots/ColorRule.js");
var import_Annotations = require("../slots/Annotations/Annotations.js");
var import_AreaSeries = require("../slots/AreaSeries.js");
var import_BarSeries = require("../slots/BarSeries.js");
var import_CustomTooltip = require("../slots/CustomTooltip.js");
var import_DotSeries = require("../slots/DotSeries.js");
var import_Legend = require("../slots/Legend.js");
var import_LineSeries = require("../slots/LineSeries.js");
var import_RectSeries = require("../slots/RectSeries.js");
var import_Select = require("../slots/Select.js");
var import_ThresholdIndicator = require("../slots/ThresholdIndicator.js");
var import_Toolbar = require("../slots/Toolbar.js");
var import_Tooltip = require("../slots/Tooltip.js");
var import_XAxis = require("../slots/XAxis.js");
var import_YAxis = require("../slots/YAxis.js");
var import_xy_chart_internals = require("../types/xy-chart-internals.js");
const iterateSlots = (children) => {
  const slots = {
    series: [],
    yAxes: [],
    xAxes: [],
    legend: {},
    toolbar: {},
    thresholds: [],
    colorRules: [],
    tooltip: void 0
  };
  let intents;
  import_react.Children.forEach(children, (child) => {
    if ((0, import_is_slot.isSlot)(child, import_RectSeries.RectSeries)) {
      slots.series.push({
        ...child.props,
        id: crypto.randomUUID(),
        type: import_xy_chart_internals.RECT_VARIANT
      });
    } else if ((0, import_is_slot.isSlot)(child, import_DotSeries.DotSeries)) {
      slots.series.push({
        ...child.props,
        id: crypto.randomUUID(),
        type: import_xy_chart_internals.DOT_VARIANT
      });
    } else if ((0, import_is_slot.isSlot)(child, import_LineSeries.LineSeries)) {
      slots.series.push({
        ...child.props,
        id: crypto.randomUUID(),
        type: import_xy_chart_internals.LINE_VARIANT
      });
    } else if ((0, import_is_slot.isSlot)(child, import_BarSeries.BarSeries)) {
      slots.series.push({
        ...child.props,
        id: crypto.randomUUID(),
        type: import_xy_chart_internals.BAR_VARIANT
      });
    } else if ((0, import_is_slot.isSlot)(child, import_AreaSeries.AreaSeries)) {
      slots.series.push({
        ...child.props,
        id: crypto.randomUUID(),
        type: import_xy_chart_internals.AREA_VARIANT
      });
    } else if ((0, import_is_slot.isSlot)(child, import_XAxis.XAxis)) {
      slots.xAxes = [...slots.xAxes, child.props];
    } else if ((0, import_is_slot.isSlot)(child, import_YAxis.YAxis)) {
      slots.yAxes = [...slots.yAxes, child.props];
    } else if ((0, import_is_slot.isSlot)(child, import_Legend.Legend)) {
      slots.legend = child.props;
    } else if ((0, import_is_slot.isSlot)(child, import_Toolbar.Toolbar)) {
      slots.toolbar = child.props;
    } else if ((0, import_is_slot.isSlot)(child, import_ThresholdIndicator.ThresholdIndicator)) {
      slots.thresholds.push(child.props);
    } else if ((0, import_is_slot.isSlot)(child, import_Intent.Intent)) {
      intents = intents ? [...intents, child.props] : [child.props];
    } else if ((0, import_is_slot.isSlot)(child, import_Select.Select)) {
      const select = child.props;
      Object.assign(slots, { select });
    } else if ((0, import_is_slot.isSlot)(child, import_Annotations.Annotations)) {
      const annotations = (0, import_iterate_xy_chart_annotations.iterateXYChartAnnotations)(
        child.props
      );
      if (annotations) {
        slots.annotations = annotations;
      }
    } else if ((0, import_is_slot.isSlot)(child, import_ColorRule.ColorRule)) {
      (0, import_ColorRule.isValidColorRule)(child.props) && slots.colorRules.push(child.props);
    } else if ((0, import_is_slot.isSlot)(child, import_CustomTooltip.CustomTooltip)) {
      slots.tooltip = {
        hidden: false,
        template: child.props.children
      };
    } else if ((0, import_is_slot.isSlot)(child, import_Tooltip.XYChartTooltipSlot)) {
      slots.tooltip = {
        hidden: child.props.hidden ?? false,
        template: child.props.children
      };
    }
  });
  return {
    ...slots,
    toolbar: {
      ...slots.toolbar,
      intents
    }
  };
};
