// @dynatrace/strato-components/charts/xy-chart/utils/query-closest-quadtree.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var query_closest_quadtree_exports = {};
__export(query_closest_quadtree_exports, {
  buildQuadtrees: () => buildQuadtrees,
  buildQueryClosest: () => buildQueryClosest,
  cancelDebouncedQuadtrees: () => cancelDebouncedQuadtrees,
  distanceBetweenTwoPoints: () => distanceBetweenTwoPoints,
  getDebouncedQuadtrees: () => getDebouncedQuadtrees
});
module.exports = __toCommonJS(query_closest_quadtree_exports);
var import_d3_quadtree = require("d3-quadtree");
var import_lodash_es = require("lodash");
var import_get_linear_accessors = require("./get-linear-accessors.js");
var import_get_scale = require("./get-scale.js");
var import_value_representation = require("./value-representation.js");
var import_constants = require("../../core/constants.js");
var import_safe_get = require("../../core/utils/safe-get.js");
function getAxisAccessors(series, metadata, xScale, yScale, axisIdToScaleDomain, datapointToMetadata) {
  const { xAxis, yAxis } = metadata;
  const xRelAccessor = (0, import_value_representation.relAccessor)(xAxis, axisIdToScaleDomain);
  const yRelAccessor = (0, import_value_representation.relAccessor)(yAxis, axisIdToScaleDomain);
  const { xAccessor, yAccessor } = (0, import_get_linear_accessors.getLinearAccessors)(
    series.type,
    metadata,
    datapointToMetadata,
    xRelAccessor,
    yRelAccessor
  );
  return [xAccessor(xScale), yAccessor(yScale)];
}
function filterDatapointsInRange(points, xAxisAccessor, yAxisAccessor, bounds, datapointToMetadata) {
  const { xMin, xMax, yMin, yMax } = bounds;
  return points.filter((point) => {
    if ((0, import_safe_get.safeGet)(datapointToMetadata, point).isGap) {
      return false;
    }
    const xValue = xAxisAccessor(point);
    const yValue = yAxisAccessor(point);
    if (xValue < xMin || xValue > xMax) {
      return false;
    }
    if (yValue < yMin || yValue > yMax) {
      return false;
    }
    return true;
  });
}
function buildSingleQuadtree(xAxisAccessor, yAxisAccessor, filtered) {
  return (0, import_d3_quadtree.quadtree)().x((d) => xAxisAccessor(d)).y((d) => yAxisAccessor(d)).addAll(filtered);
}
function buildQuadtrees(seriesWithDatapoints, axisIdToScales, datapointToMetadata, seriesToMetadata, axisIdToScaleDomain) {
  const seriesToTrees = /* @__PURE__ */ new Map();
  for (const series of seriesWithDatapoints) {
    const xScale = (0, import_get_scale.getCurrentScale)(axisIdToScales, series.xAxisId);
    const yScale = (0, import_get_scale.getCurrentScale)(axisIdToScales, series.yAxisId);
    const metadata = (0, import_safe_get.safeGet)(seriesToMetadata, series);
    const accessors = getAxisAccessors(
      series,
      metadata,
      xScale,
      yScale,
      axisIdToScaleDomain,
      datapointToMetadata
    );
    const [xAxisAccessor, yAxisAccessor] = accessors;
    const [xMin, xMax] = xScale.range();
    const [yMin, yMax] = yScale.range();
    const filtered = filterDatapointsInRange(
      series.data,
      xAxisAccessor,
      yAxisAccessor,
      {
        xMin: Math.min(xMin, xMax),
        xMax: Math.max(xMin, xMax),
        yMin: Math.min(yMin, yMax),
        yMax: Math.max(yMin, yMax)
      },
      datapointToMetadata
    );
    const tree = buildSingleQuadtree(xAxisAccessor, yAxisAccessor, filtered);
    seriesToTrees.set(series.id, tree);
  }
  return seriesToTrees;
}
const DELAY = 300;
const debounceMap = /* @__PURE__ */ new Map();
function getDebouncedQuadtrees(key) {
  let debounced = debounceMap.get(key);
  if (!debounced) {
    debounced = (0, import_lodash_es.debounce)(
      (series, scales, metadata, seriesToMetadata, axisIdToScaleDomain, onDone) => onDone(
        buildQuadtrees(
          series,
          scales,
          metadata,
          seriesToMetadata,
          axisIdToScaleDomain
        )
      ),
      DELAY
    );
    debounceMap.set(key, debounced);
  }
  return debounced;
}
function cancelDebouncedQuadtrees(key) {
  debounceMap.get(key)?.cancel?.();
}
function distanceBetweenTwoPoints(x0, x1, y0, y1) {
  return Math.sqrt(Math.pow(x0 - x1, 2) + Math.pow(y0 - y1, 2));
}
function getDatapointAxisValues(datapoint, series, metadata, xScale, yScale, axisIdToScaleDomain, datapointToMetadata) {
  const accessors = getAxisAccessors(
    series,
    metadata,
    xScale,
    yScale,
    axisIdToScaleDomain,
    datapointToMetadata
  );
  return [accessors[0](datapoint), accessors[1](datapoint)];
}
function findClosestDatapointInSeries(series, tree, position, metadata, queryContext) {
  const datapoint = tree.find(position.x, position.y);
  if (!datapoint) {
    return void 0;
  }
  const [xValue, yValue] = getDatapointAxisValues(
    datapoint,
    series,
    metadata,
    queryContext.xScale,
    queryContext.yScale,
    queryContext.axisIdToScaleDomain,
    queryContext.datapointToMetadata
  );
  const distance = distanceBetweenTwoPoints(
    position.x,
    xValue,
    position.y,
    yValue
  );
  return {
    type: series.type,
    distance,
    datapoint,
    series
  };
}
function buildQueryClosest(seriesIdToQuadtree, axisIdToScales, seriesToMetadata, axisIdToScaleDomain, datapointToMetadata) {
  function query(arrayOfSeries, position) {
    let best;
    for (const series of arrayOfSeries) {
      const tree = seriesIdToQuadtree.get(series.id);
      if (!tree) {
        continue;
      }
      const xScale = (0, import_get_scale.getCurrentScale)(axisIdToScales, series.xAxisId);
      const yScale = (0, import_get_scale.getCurrentScale)(axisIdToScales, series.yAxisId);
      const metadata = (0, import_safe_get.safeGet)(seriesToMetadata, series);
      const closest = findClosestDatapointInSeries(
        series,
        tree,
        position,
        metadata,
        {
          xScale,
          yScale,
          axisIdToScaleDomain,
          datapointToMetadata
        }
      );
      if (!closest) {
        continue;
      }
      if (!best || closest.distance < best.distance) {
        best = closest;
      }
      if (best && best.distance <= import_constants.DEFAULT_DATA_POINT_RADIUS) {
        break;
      }
    }
    return best;
  }
  return query;
}
