// @dynatrace/strato-components/charts/xy-chart/utils/query-hovered.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var query_hovered_exports = {};
__export(query_hovered_exports, {
  buildQueryHovered: () => buildQueryHovered,
  getHoveredDatapoint: () => getHoveredDatapoint,
  getPixelDimensionsForDatapoint: () => getPixelDimensionsForDatapoint
});
module.exports = __toCommonJS(query_hovered_exports);
var import_get_scale = require("./get-scale.js");
var import_type_guards = require("./type-guards.js");
var import_value_representation = require("./value-representation.js");
var import_safe_get = require("../../core/utils/safe-get.js");
var import_bar_accessors = require("../components/displays/bar-accessors.js");
var import_rect_accessors = require("../components/displays/rect-accessors.js");
var import_xy_chart_internals = require("../types/xy-chart-internals.js");
function isInPixelRange(position, start, span) {
  return position >= start && position <= start + span;
}
function getAccessors(series, axisIdToScales, datapointMetadataMap, xRelAccessor, yRelAccessor) {
  const { type, xAxisId, yAxisId } = series;
  const xScale = (0, import_get_scale.getCurrentScale)(axisIdToScales, xAxisId, type);
  const yScale = (0, import_get_scale.getCurrentScale)(axisIdToScales, yAxisId, type);
  if (type === import_xy_chart_internals.RECT_VARIANT) {
    return {
      xAccessor: (0, import_rect_accessors.xAccessor)(xScale, xRelAccessor),
      yAccessor: (0, import_rect_accessors.yAccessor)(yScale, yRelAccessor),
      widthAccessor: (0, import_rect_accessors.widthAccessor)(xScale, xRelAccessor),
      heightAccessor: (0, import_rect_accessors.heightAccessor)(yScale, yRelAccessor)
    };
  }
  const dimensionXScale = (0, import_get_scale.getDimensionScale)(axisIdToScales, xAxisId);
  const dimensionYScale = (0, import_get_scale.getDimensionScale)(axisIdToScales, yAxisId);
  return {
    xAccessor: (0, import_bar_accessors.xAccessor)(
      xScale,
      datapointMetadataMap,
      dimensionXScale,
      series.id,
      xRelAccessor
    ),
    yAccessor: (0, import_bar_accessors.yAccessor)(
      yScale,
      datapointMetadataMap,
      dimensionYScale,
      series.id,
      yRelAccessor
    ),
    widthAccessor: (0, import_bar_accessors.widthAccessor)(
      xScale,
      datapointMetadataMap,
      dimensionXScale,
      xRelAccessor
    ),
    heightAccessor: (0, import_bar_accessors.heightAccessor)(
      yScale,
      datapointMetadataMap,
      dimensionYScale,
      yRelAccessor
    )
  };
}
const getPixelDimensionsForDatapoint = (datapoint, genAccessors) => {
  if ((0, import_type_guards.isRectDatapoint)(datapoint)) {
    const accessors2 = genAccessors;
    return {
      x: accessors2.xAccessor(datapoint),
      y: accessors2.yAccessor(datapoint),
      width: accessors2.widthAccessor(datapoint),
      height: accessors2.heightAccessor(datapoint)
    };
  }
  const accessors = genAccessors;
  return {
    x: accessors.xAccessor(datapoint),
    y: accessors.yAccessor(datapoint),
    width: accessors.widthAccessor(datapoint),
    height: accessors.heightAccessor(datapoint)
  };
};
function getHoveredDatapoint(position, series, axisIdToScales, datapointMetadataMap, seriesToMetadata, axisIdToScaleDomain) {
  const data = (0, import_type_guards.isRectSeries)(series) ? series.filteredData : series.data;
  if (position.x < 0 || position.y < 0) {
    return void 0;
  }
  const { xAxis, yAxis } = (0, import_safe_get.safeGet)(seriesToMetadata, series);
  const xRelAccessor = (0, import_value_representation.relAccessor)(xAxis, axisIdToScaleDomain);
  const yRelAccessor = (0, import_value_representation.relAccessor)(yAxis, axisIdToScaleDomain);
  const accessors = getAccessors(
    series,
    axisIdToScales,
    datapointMetadataMap,
    xRelAccessor,
    yRelAccessor
  );
  for (const datapoint of data) {
    if ((0, import_safe_get.safeGet)(datapointMetadataMap, datapoint).isGap) {
      continue;
    }
    const { x, y, width, height } = getPixelDimensionsForDatapoint(
      datapoint,
      accessors
    );
    if (isInPixelRange(position.x, x, width) && isInPixelRange(position.y, y, height)) {
      return datapoint;
    }
  }
  return void 0;
}
function buildQueryHovered(axisIdToScales, datapointMetadataMap, seriesToMetadata, axisIdToScaleDomain) {
  function query(arrayOfSeries, position) {
    for (let i = arrayOfSeries.length - 1; i >= 0; i--) {
      const series = arrayOfSeries[i];
      const datapoint = getHoveredDatapoint(
        position,
        series,
        axisIdToScales,
        datapointMetadataMap,
        seriesToMetadata,
        axisIdToScaleDomain
      );
      if (!datapoint) {
        continue;
      }
      return {
        type: series.type,
        distance: 0,
        datapoint,
        series
      };
    }
    return void 0;
  }
  return query;
}
