// @dynatrace/strato-components/charts/xy-chart/utils/query-get-closest-point.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var query_get_closest_point_exports = {};
__export(query_get_closest_point_exports, {
  buildQueryGetClosestPoint: () => buildQueryGetClosestPoint,
  getClosestDatapoint: () => getClosestDatapoint,
  getDistance: () => getDistance
});
module.exports = __toCommonJS(query_get_closest_point_exports);
var import_lodash_es = require("lodash");
var import_get_linear_accessors = require("./get-linear-accessors.js");
var import_get_scale = require("./get-scale.js");
var import_query_closest_quadtree = require("./query-closest-quadtree.js");
var import_value_representation = require("./value-representation.js");
var import_constants = require("../../core/constants.js");
var import_detect_scale = require("../../core/utils/detect-scale.js");
var import_safe_get = require("../../core/utils/safe-get.js");
function getDistance(datapoint, xScale, yScale, relativePosition, xAccessor, yAccessor) {
  const xPixel = xAccessor(xScale)(datapoint);
  const yPixel = yAccessor(yScale)(datapoint);
  return (0, import_query_closest_quadtree.distanceBetweenTwoPoints)(
    relativePosition.x,
    xPixel,
    relativePosition.y,
    yPixel
  );
}
function calculateScalesBounds(xScale, yScale) {
  const [minWidth, maxWidth] = xScale.range();
  const yScaleRange = yScale.range();
  const [minHeight, maxHeight] = (0, import_detect_scale.isBandScale)(yScale) ? [yScaleRange[0], yScaleRange[1]] : [yScaleRange[1], yScaleRange[0]];
  return { minWidth, maxWidth, minHeight, maxHeight };
}
function getAccessorsForSeries(seriesType, seriesToMetadata, currentSeries, datapointToMetadata, xRelAccessor, yRelAccessor) {
  const metadata = (0, import_safe_get.safeGet)(seriesToMetadata, currentSeries);
  return (0, import_get_linear_accessors.getLinearAccessors)(
    seriesType,
    metadata,
    datapointToMetadata,
    xRelAccessor,
    yRelAccessor
  );
}
function isDatapointOutOfBounds(xPixel, yPixel, bounds) {
  const { minWidth, maxWidth, minHeight, maxHeight } = bounds;
  return xPixel < minWidth || xPixel > maxWidth || yPixel < minHeight || yPixel > maxHeight;
}
function findClosestInSeriesData(data, xScale, yScale, relativePosition, accessors, bounds, currentAccumulator) {
  let bestInSeries = currentAccumulator;
  let datapointIndex = 0;
  while (datapointIndex < data.length) {
    const datapoint = data[datapointIndex];
    const xPixel = accessors.xAccessor(xScale)(datapoint);
    const yPixel = accessors.yAccessor(yScale)(datapoint);
    if (isDatapointOutOfBounds(xPixel, yPixel, bounds)) {
      datapointIndex += 1;
      continue;
    }
    const distance = getDistance(
      datapoint,
      xScale,
      yScale,
      relativePosition,
      accessors.xAccessor,
      accessors.yAccessor
    );
    if (distance < bestInSeries.distance) {
      bestInSeries = {
        distance,
        datapoint,
        series: currentAccumulator.series
      };
    }
    if (distance <= import_constants.DEFAULT_DATA_POINT_RADIUS) {
      return bestInSeries;
    }
    datapointIndex += 1;
  }
  return bestInSeries.distance < currentAccumulator.distance ? bestInSeries : void 0;
}
function getClosestDatapoint(relativePosition, series, axisIdToScales, seriesToMetadata, datapointToMetadata, axisIdToScaleDomain) {
  if (!series.length) {
    return void 0;
  }
  const initialAccumulator = {
    distance: Infinity,
    datapoint: {},
    series: {}
  };
  let bestMatch = initialAccumulator;
  let seriesIndex = series.length - 1;
  while (seriesIndex >= 0) {
    const currentSeries = series[seriesIndex];
    const { data, xAxisId, yAxisId, type } = currentSeries;
    const xScale = (0, import_get_scale.getCurrentScale)(axisIdToScales, xAxisId);
    const yScale = (0, import_get_scale.getCurrentScale)(axisIdToScales, yAxisId);
    const seriesMetadata = (0, import_safe_get.safeGet)(seriesToMetadata, currentSeries);
    const xRelAccessor = axisIdToScaleDomain ? (0, import_value_representation.relAccessor)(seriesMetadata.xAxis, axisIdToScaleDomain) : void 0;
    const yRelAccessor = axisIdToScaleDomain ? (0, import_value_representation.relAccessor)(seriesMetadata.yAxis, axisIdToScaleDomain) : void 0;
    const accessors = getAccessorsForSeries(
      type,
      seriesToMetadata,
      currentSeries,
      datapointToMetadata,
      xRelAccessor,
      yRelAccessor
    );
    const bounds = calculateScalesBounds(xScale, yScale);
    const seriesAccumulator = {
      ...initialAccumulator,
      series: currentSeries
    };
    const updatedAccumulator = findClosestInSeriesData(
      data,
      xScale,
      yScale,
      relativePosition,
      accessors,
      bounds,
      seriesAccumulator
    );
    if (updatedAccumulator) {
      bestMatch = updatedAccumulator;
      if (updatedAccumulator.distance <= import_constants.DEFAULT_DATA_POINT_RADIUS) {
        break;
      }
    }
    seriesIndex -= 1;
  }
  if ((0, import_lodash_es.isEmpty)(bestMatch.datapoint) || (0, import_lodash_es.isEmpty)(bestMatch.series)) {
    return void 0;
  }
  return {
    datapoint: bestMatch.datapoint,
    series: bestMatch.series,
    distance: bestMatch.distance
  };
}
function buildQueryGetClosestPoint(axisIdToScales, seriesToMetadata, datapointToMetadata, axisIdToScaleDomain) {
  function query(arrayOfSeries, position) {
    if ((0, import_lodash_es.isEmpty)(arrayOfSeries)) {
      return void 0;
    }
    const closest = getClosestDatapoint(
      position,
      arrayOfSeries,
      axisIdToScales,
      seriesToMetadata,
      datapointToMetadata,
      axisIdToScaleDomain
    );
    if (!closest) {
      return void 0;
    }
    return {
      type: closest.series.type,
      distance: closest?.distance,
      datapoint: closest?.datapoint,
      series: closest?.series
    };
  }
  return query;
}
