// @dynatrace/strato-components/charts/xy-chart/utils/type-guards.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var type_guards_exports = {};
__export(type_guards_exports, {
  isAreaSeries: () => isAreaSeries,
  isAreaTooltipContent: () => isAreaTooltipContent,
  isAreaTooltipContext: () => isAreaTooltipContext,
  isBarSeries: () => isBarSeries,
  isBarTooltipContext: () => isBarTooltipContext,
  isCategoricalAreaScale: () => isCategoricalAreaScale,
  isCategoricalBarScale: () => isCategoricalBarScale,
  isCategoricalDotScale: () => isCategoricalDotScale,
  isCategoricalHoveringPoint: () => isCategoricalHoveringPoint,
  isCategoricalLineScale: () => isCategoricalLineScale,
  isCategoricalRectScale: () => isCategoricalRectScale,
  isCategoricalTooltip: () => isCategoricalTooltip,
  isDotSeries: () => isDotSeries,
  isDotTooltipContent: () => isDotTooltipContent,
  isHorizontalAreaAccessors: () => isHorizontalAreaAccessors,
  isLineSeries: () => isLineSeries,
  isLineTooltipContent: () => isLineTooltipContent,
  isNumericalAreaScale: () => isNumericalAreaScale,
  isNumericalBarScale: () => isNumericalBarScale,
  isNumericalDotScale: () => isNumericalDotScale,
  isNumericalHoveringPoint: () => isNumericalHoveringPoint,
  isNumericalLineScale: () => isNumericalLineScale,
  isNumericalRectScale: () => isNumericalRectScale,
  isNumericalToolTip: () => isNumericalToolTip,
  isRectDatapoint: () => isRectDatapoint,
  isRectSeries: () => isRectSeries,
  isRectTooltipContext: () => isRectTooltipContext,
  isStringOrValidNumber: () => isStringOrValidNumber,
  isTimeAreaScale: () => isTimeAreaScale,
  isTimeBarScale: () => isTimeBarScale,
  isTimeDotScale: () => isTimeDotScale,
  isTimeHoveringPoint: () => isTimeHoveringPoint,
  isTimeLineScale: () => isTimeLineScale,
  isTimeRectScale: () => isTimeRectScale,
  isTimeTooltip: () => isTimeTooltip,
  isValidName: () => isValidName,
  isVerticalAreaAccessors: () => isVerticalAreaAccessors
});
module.exports = __toCommonJS(type_guards_exports);
var import_xy_chart_internals = require("../types/xy-chart-internals.js");
function isNumerical(args) {
  return typeof args.v0 === "number";
}
function isCategorical(args) {
  return typeof args.v0 === "string";
}
function isTime(args) {
  return args.v0 instanceof Date;
}
function isNumericalHoveringPoint(args) {
  return isNumerical(args);
}
function isTimeHoveringPoint(args) {
  return isTime(args);
}
function isCategoricalHoveringPoint(args) {
  return isCategorical(args);
}
function isNumericalDotScale(args) {
  return isNumerical(args);
}
function isTimeDotScale(args) {
  return isTime(args);
}
function isCategoricalDotScale(args) {
  return isCategorical(args);
}
function isNumericalRectScale(args) {
  return isNumerical(args);
}
function isTimeRectScale(args) {
  return isTime(args);
}
function isCategoricalRectScale(args) {
  return isCategorical(args);
}
function isNumericalLineScale(args) {
  return isNumerical(args);
}
function isTimeLineScale(args) {
  return isTime(args);
}
function isCategoricalLineScale(args) {
  return isCategorical(args);
}
function isNumericalBarScale(args) {
  return typeof args.v0 === "number" || typeof args.v1 === "number";
}
function isTimeBarScale(args) {
  return args.v0 instanceof Date || args.v1 instanceof Date;
}
function isCategoricalBarScale(args) {
  return typeof args.v0 === "string" || typeof args.v1 === "string";
}
function isNumericalAreaScale(args) {
  return typeof args.v0 === "number" || typeof args.v1 === "number";
}
function isTimeAreaScale(args) {
  return args.v0 instanceof Date || args.v1 instanceof Date;
}
function isCategoricalAreaScale(args) {
  return typeof args.v0 === "string" || typeof args.v1 === "string";
}
function isNumericalToolTip(args) {
  return isNumerical(args);
}
function isTimeTooltip(args) {
  return isTime(args);
}
function isCategoricalTooltip(args) {
  return isCategorical(args);
}
function isRectSeries(series) {
  return series.type === import_xy_chart_internals.RECT_VARIANT;
}
function isDotSeries(series) {
  return series.type === import_xy_chart_internals.DOT_VARIANT;
}
function isLineSeries(series) {
  return series.type === import_xy_chart_internals.LINE_VARIANT;
}
function isBarSeries(series) {
  return series.type === import_xy_chart_internals.BAR_VARIANT;
}
function isAreaSeries(series) {
  return series.type === import_xy_chart_internals.AREA_VARIANT;
}
function isRectTooltipContext(context) {
  return context.series.type === "rect" && !!context.metadata;
}
function isBarTooltipContext(context) {
  return context.series.type === "bar" && !!context.metadata;
}
function isAreaTooltipContext(context) {
  return context.series.type === "area" && !!context.metadata;
}
function isDotTooltipContent(content) {
  return content?.series?.type === import_xy_chart_internals.DOT_VARIANT;
}
function isLineTooltipContent(content) {
  return content?.series?.type === import_xy_chart_internals.LINE_VARIANT;
}
function isAreaTooltipContent(content) {
  return content?.series?.type === import_xy_chart_internals.AREA_VARIANT;
}
function isRectDatapoint(datapoint) {
  return "value" in datapoint;
}
const isStringOrValidNumber = (value) => {
  return typeof value === "string" || typeof value === "number" && !Number.isNaN(value);
};
function isValidName(value) {
  return value !== null && value !== void 0 && !Number.isNaN(value);
}
function isHorizontalAreaAccessors(accessors) {
  return accessors.x.length === 4;
}
function isVerticalAreaAccessors(accessors) {
  return accessors.y.length === 4;
}
