// @dynatrace/strato-components/charts/xy-chart/utils/xy-chart-configuration.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var xy_chart_configuration_exports = {};
__export(xy_chart_configuration_exports, {
  buildInternalConfig: () => buildInternalConfig,
  createXYChartConfiguration: () => createXYChartConfiguration,
  deserializeConfig: () => deserializeConfig,
  getAxes: () => getAxes,
  getTruncationModeConfiguration: () => getTruncationModeConfiguration,
  getXYChartAnnotationsConfiguration: () => getXYChartAnnotationsConfiguration
});
module.exports = __toCommonJS(xy_chart_configuration_exports);
var import_lodash_es = require("lodash");
var import_axis_interactions = require("./axis-interactions.js");
var import_defaults = require("./defaults.js");
var import_is_axis = require("./is-axis.js");
var import_constants = require("../../core/constants.js");
var import_ColorRule = require("../../core/slots/ColorRule.js");
var import_color_palette = require("../../core/types/color-palette.js");
var import_colored_range = require("../../core/types/colored-range.js");
var import_is_valid_color_palette = require("../../core/utils/is-valid-color-palette.js");
var import_is_valid_legend_position = require("../../core/utils/is-valid-legend-position.js");
var import_is_valid_toolbar_placement = require("../../core/utils/is-valid-toolbar-placement.js");
var import_builder = require("../components/thresholds/builder.js");
var import_defaults2 = require("../constants/defaults.js");
var import_XYChartCustomTooltip_context = require("../context/XYChartCustomTooltip.context.js");
var import_axis = require("../types/axis.js");
var import_xy_chart = require("../types/xy-chart.js");
function getLegendConfiguration(slotsLegend, contextLegend) {
  const legendConfig = (0, import_lodash_es.merge)(
    {},
    import_defaults2.defaultLegend,
    contextLegend,
    slotsLegend
  );
  legendConfig.position = (0, import_is_valid_legend_position.isValidLegendPosition)(legendConfig.position) ? legendConfig.position : import_defaults2.defaultLegend.position;
  return legendConfig;
}
function getToolbarConfiguration(slotsToolbar, contextToolbar) {
  const toolbarConfig = (0, import_lodash_es.merge)({}, import_defaults2.defaultToolbar, contextToolbar, slotsToolbar);
  toolbarConfig.placement = (0, import_is_valid_toolbar_placement.isValidToolbarPlacement)(toolbarConfig.placement) ? toolbarConfig.placement : import_defaults2.defaultToolbar.placement;
  return toolbarConfig;
}
function getAxesWithSeries(axisSlots, series) {
  let axesConfig = [];
  const matchingIds = /* @__PURE__ */ new Set();
  axisSlots.forEach((axis) => {
    const isAlreadyDefined = matchingIds.has(axis.id);
    if (isAlreadyDefined) {
      return;
    }
    const hasMatchingSeriesId = series.find((series2) => {
      return (0, import_is_axis.isXAxis)(axis) ? series2.xAxisId === axis.id : series2.yAxisId === axis.id;
    });
    if (hasMatchingSeriesId) {
      matchingIds.add(axis.id);
      axesConfig = [...axesConfig, axis];
    }
  });
  return axesConfig;
}
function getColorPaletteConfiguration(contextColorPalette, propsColorPalette) {
  const colorPalette = propsColorPalette ?? contextColorPalette ?? void 0;
  if (!colorPalette) {
    return void 0;
  }
  return (0, import_colored_range.isColoredRangeArray)(colorPalette) || (0, import_color_palette.isCustomColorPalette)(colorPalette) || (0, import_is_valid_color_palette.isValidColorPalette)(colorPalette) ? colorPalette : void 0;
}
function getTruncationModeConfiguration(contextTruncationMode, propsTruncationMode) {
  return propsTruncationMode ?? contextTruncationMode ?? import_constants.TRUNCATION_MODE.Middle;
}
function getXYChartAnnotationsConfiguration(slotsAnnotations, configAnnotations) {
  if ((0, import_lodash_es.isEmpty)(slotsAnnotations) && !configAnnotations) {
    return void 0;
  }
  return (0, import_lodash_es.merge)({}, configAnnotations, slotsAnnotations);
}
function getCustomTooltipConfiguration(slotsCustomTooltip, contextCustomTooltip) {
  if (slotsCustomTooltip !== void 0) {
    return slotsCustomTooltip;
  }
  return {
    ...import_XYChartCustomTooltip_context.DEFAULT_CUSTOM_TOOLTIP_CONFIG,
    hidden: contextCustomTooltip?.hidden ?? import_XYChartCustomTooltip_context.DEFAULT_CUSTOM_TOOLTIP_CONFIG.hidden
  };
}
const createXYChartConfiguration = (slots, contextConfig, rootProps) => {
  const colorPalette = getColorPaletteConfiguration(
    contextConfig.colorPalette,
    rootProps?.colorPalette
  );
  const truncationMode = getTruncationModeConfiguration(
    contextConfig.truncationMode,
    rootProps?.truncationMode
  );
  const legend = getLegendConfiguration(slots.legend, contextConfig.legend);
  const xAxes = getAxes(slots.xAxes, contextConfig.xAxes);
  const yAxes = getAxes(slots.yAxes, contextConfig.yAxes);
  const toolbar = getToolbarConfiguration(slots.toolbar, contextConfig.toolbar);
  const annotations = getXYChartAnnotationsConfiguration(
    slots.annotations,
    contextConfig.annotations
  );
  const select = slots.select || void 0;
  const thresholdIndicators = [
    ...slots.thresholds,
    ...contextConfig.thresholdIndicators || []
  ];
  const colorRules = (contextConfig.colorRules || []).filter(import_ColorRule.isValidColorRule).concat(slots.colorRules || []);
  const tooltip = getCustomTooltipConfiguration(
    slots.tooltip,
    contextConfig.tooltip
  );
  return {
    colorPalette,
    legend,
    toolbar,
    truncationMode,
    xAxes,
    yAxes,
    thresholdIndicators,
    annotations,
    select,
    colorRules,
    tooltip,
    customTooltip: {}
  };
};
function getAxes(slotAxes, contextAxes) {
  const configContextAxes = contextAxes ?? [];
  const ids = /* @__PURE__ */ new Set();
  const axesConfig = [...slotAxes, ...configContextAxes];
  return axesConfig.reduce((acc, axis) => {
    if (!ids.has(axis.id)) {
      acc.push({ ...import_defaults2.DEFAULT_AXIS_CONFIG, ...axis });
      ids.add(axis.id);
    }
    return acc;
  }, []);
}
const deserializeConfig = (serializedConfig) => {
  try {
    return (0, import_lodash_es.merge)({}, import_defaults2.defaultConfig, JSON.parse(serializedConfig));
  } catch {
    return import_defaults2.defaultConfig;
  }
};
const buildInternalConfig = (config, slots) => {
  const yAxes = config.yAxes.map((axis) => ({
    ...axis,
    coordinate: "y",
    tickLabelLayout: "horizontal",
    type: import_axis.XYAxisMapper[axis.type],
    interactions: { zoom: false, pan: false, infinite: false },
    onZoomChange: import_defaults.doesNothing,
    isRelative: axis.type === import_xy_chart.NUMERICAL_SCALE_TYPE && axis.valueRepresentation === "relative"
  }));
  const xAxes = config.xAxes.map((axis) => ({
    ...axis,
    coordinate: "x",
    tickLabelLayout: axis.tickLabelLayout ?? "horizontal",
    type: import_axis.XYAxisMapper[axis.type],
    interactions: (0, import_axis_interactions.interactionsFor)(axis),
    onZoomChange: axis.onZoomChange ?? import_defaults.doesNothing,
    isRelative: axis.type === import_xy_chart.NUMERICAL_SCALE_TYPE && axis.valueRepresentation === "relative"
  }));
  const axes = getAxesWithSeries(
    [...yAxes, ...xAxes],
    slots.series
  );
  const thresholds = (0, import_builder.buildThresholds)(yAxes, xAxes, config.thresholdIndicators);
  return { series: slots.series, axes, thresholds };
};
