// @dynatrace/strato-components/esm/charts/gantt-chart/components/tooltip/GanttChartTooltip.js
import { jsx } from "react/jsx-runtime";
import { useContext, useRef } from "react";
import { ganttChartTooltipCSS } from "./GanttChartTooltip.sty.js";
import { GanttChartTooltipContent } from "./GanttChartTooltipContent.js";
import { useChartActionsMenu } from "../../../core/components/series-action/hooks/useChartActionsMenu.js";
import { ChartTooltip } from "../../../core/components/tooltip/ChartTooltip/ChartTooltip.js";
import { DEFAULT_TOOLTIP_MARGIN } from "../../constants/defaults.js";
import { GanttSegmentsContext } from "../../context/GanttSegmentsContext.js";
import { useRegisterTooltipHandlerCallbacks } from "../../hooks/use-register-tooltip-handler-callbacks.js";
import {
  useTooltipPinnedStatus,
  useTooltipPosition
} from "../../hooks/use-set-tooltip-state.js";
const GanttChartTooltip = () => {
  const tooltipRef = useRef(null);
  const pinned = useTooltipPinnedStatus();
  const mousePosition = useTooltipPosition();
  const { hovering = true } = useContext(GanttSegmentsContext);
  const {
    selectedItem: selectedItemId,
    actionsMenuRef,
    updateSelectedItem,
    onLeave
  } = useChartActionsMenu();
  useRegisterTooltipHandlerCallbacks(tooltipRef);
  const isTooltipHeightKnown = useRef(false);
  if (!mousePosition || !hovering) {
    return null;
  }
  if (!isTooltipHeightKnown.current && tooltipRef.current) {
    isTooltipHeightKnown.current = true;
    return null;
  }
  const position = getTooltipPosition(mousePosition, tooltipRef);
  return /* @__PURE__ */ jsx(
    ChartTooltip,
    {
      ref: tooltipRef,
      pinned,
      className: ganttChartTooltipCSS,
      position,
      onMouseLeave: onLeave,
      actionSelectionValue: {
        selectedItemId,
        actionsMenuRef,
        updateSelectedItem
      },
      children: /* @__PURE__ */ jsx(GanttChartTooltipContent, {})
    }
  );
};
function getTooltipPosition(mousePosition, tooltipRef) {
  if (!tooltipRef.current) {
    return {
      x: -9999,
      y: -9999
    };
  }
  const position = {
    x: mousePosition.x,
    y: mousePosition.y + DEFAULT_TOOLTIP_MARGIN
  };
  const tooltipHeight = tooltipRef.current.getBoundingClientRect().height;
  const viewportHeight = window.innerHeight;
  if (position.y + tooltipHeight > viewportHeight) {
    position.y = viewportHeight - tooltipHeight;
  }
  return {
    x: Math.max(0, position.x),
    y: Math.max(0, position.y)
  };
}
export {
  GanttChartTooltip
};
