// @dynatrace/strato-components/esm/charts/gantt-chart/components/tooltip/GanttChartTooltipContent.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { useEffect, useState } from "react";
import Spacings from "@dynatrace/strato-design-tokens/spacings";
import { GanttChartTooltipBody } from "./GanttChartTooltipBody.js";
import { GanttChartTooltipHeader } from "./GanttChartTooltipHeader.js";
import { Divider } from "../../../../layouts/divider/Divider.js";
import { ChartTooltip } from "../../../core/components/tooltip/ChartTooltip/ChartTooltip.js";
import { useGanttSegmentsContext } from "../../context/GanttSegmentsContext.js";
import {
  useTooltipContent,
  useTooltipPinnedStatus
} from "../../hooks/use-set-tooltip-state.js";
const GanttChartTooltipContent = () => {
  const [tooltipContent, setTooltipContent] = useState();
  const isTooltipPinned = useTooltipPinnedStatus();
  const storedTooltipContent = useTooltipContent();
  const {
    config: { xAxisType, tooltip, formatter }
  } = useGanttSegmentsContext();
  const isDate = xAxisType === "time";
  useEffect(() => {
    if (!isTooltipPinned && storedTooltipContent) {
      setTooltipContent(storedTooltipContent);
    }
  }, [storedTooltipContent, isTooltipPinned]);
  if (!tooltipContent) {
    return null;
  }
  const { row, segment, parent } = tooltipContent;
  if (tooltip !== void 0) {
    return tooltip(segment, row, parent);
  }
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(
      GanttChartTooltipHeader,
      {
        segment: tooltipContent.segment,
        isDate,
        formatter
      }
    ),
    /* @__PURE__ */ jsx(
      Divider,
      {
        orientation: "horizontal",
        style: { margin: `${Spacings.Size4} 0` }
      }
    ),
    /* @__PURE__ */ jsx(ChartTooltip.Body, { children: /* @__PURE__ */ jsx(
      GanttChartTooltipBody,
      {
        entry: tooltipContent.segment,
        row,
        parent,
        isDate
      }
    ) })
  ] });
};
GanttChartTooltipContent["displayName"] = "GanttChartTooltipBody";
export {
  GanttChartTooltipContent
};
