// @dynatrace/strato-components/esm/charts/histogram/hooks/store/selectors.js
import { isEqual } from "lodash-es";
import { useStoreSelector } from "../../../../core/store/selector.js";
import { HistogramChartStateContext } from "../../contexts/HistogramChartState.context.js";
const useSelector = (selector, isEqual2) => useStoreSelector(HistogramChartStateContext, selector, isEqual2);
const useFilteredSeriesIdsState = () => useSelector((state) => state.filteredSeriesIds);
const useHighlightedSeriesIdState = () => useSelector((state) => state.highlightedSeriesId);
const useCurrentDomainState = () => useSelector((state) => state.currentDomain);
const useZoomState = () => useSelector((state) => state.zoom);
const useHoveredThresholdIdState = () => useSelector((state) => state.hoveredThresholdId);
const useAnnotationsState = () => useSelector((state) => state.annotations);
const useInteractionModeState = () => useSelector((state) => state.interactionMode);
const useExploreState = () => useSelector((state) => state.explore, isEqual);
const useSelectState = () => useSelector((state) => state.select, isEqual);
const useClosestBinsState = () => useSelector((state) => state.closestBins, isEqual);
export {
  useAnnotationsState,
  useClosestBinsState,
  useCurrentDomainState,
  useExploreState,
  useFilteredSeriesIdsState,
  useHighlightedSeriesIdState,
  useHoveredThresholdIdState,
  useInteractionModeState,
  useSelectState,
  useSelector,
  useZoomState
};
