// @dynatrace/strato-components/charts/categorical-bar/components/interactions/utils/data-points.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var data_points_exports = {};
__export(data_points_exports, {
  filterOutOfChartBounds: () => filterOutOfChartBounds,
  generateGroupDataPoints: () => generateGroupDataPoints,
  getGroupDataPointsBy: () => getGroupDataPointsBy,
  mapToDataPoint: () => mapToDataPoint
});
module.exports = __toCommonJS(data_points_exports);
var import_constants = require("../../../../core/components/data-point/constants.js");
var import_group_by_threshold = require("../../../../core/utils/group-by-threshold.js");
var import_layout = require("../../../utils/layout.js");
const mapToDataPoint = (dimension) => ({
  ...dimension,
  x: dimension.x + import_constants.OVERLAY_PADDING,
  y: dimension.y + import_constants.OVERLAY_PADDING,
  color: dimension.datapoint.color,
  id: dimension.datapoint.id
});
const filterOutOfChartBounds = (chartLimit, coordinate) => (dimension) => dimension[coordinate] >= import_constants.OVERLAY_PADDING && dimension[coordinate] <= chartLimit + import_constants.OVERLAY_PADDING;
function getGroupDataPointsBy(groupDataPoints, dimensionId) {
  return groupDataPoints?.filter(
    (points) => points.some((d) => d.id === dimensionId)
  );
}
function generateGroupDataPoints(points, isGroupAllowed, layout, chartLimit) {
  const coordinate = (0, import_layout.isVertical)(layout) ? "y" : "x";
  const mappedPoints = points.map(mapToDataPoint).filter(filterOutOfChartBounds(chartLimit, coordinate));
  if (!isGroupAllowed) {
    return mappedPoints.map((point) => [point]);
  }
  return (0, import_group_by_threshold.getGroupedDataPointsByCoordinate)(mappedPoints, coordinate);
}
