// @dynatrace/strato-components/esm/overlays/tooltip/TooltipElement.js
import { jsx } from "react/jsx-runtime";
import { eventListeners } from "@popperjs/core";
import { assignInlineVars } from "@vanilla-extract/dynamic";
import clsx from "clsx";
import {
  forwardRef,
  useCallback,
  useMemo,
  useRef,
  useState
} from "react";
import { usePopper } from "react-popper";
import {
  tooltipAnimationWrapperCSS,
  tooltipElementCSS,
  tooltipOffsetVar
} from "./tooltip.sty.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useMergeRefs } from "../../core/hooks/useMergeRefs.js";
import { useOverlayStacking } from "../../core/hooks/useOverlayStacking.js";
import { Text } from "../../typography/text/Text.js";
import {
  observeAnimationDone,
  observeOverlaySize
} from "../utils/popper-modifiers.js";
const TOOLTIP_OFFSET = 4;
const TooltipElement = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const {
    children,
    placement,
    triggerElement,
    id,
    fallbackPlacements: fallbackPlacementsProp,
    onMouseLeave,
    onTooltipHovered,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "data-dt-component": dataDtComponent,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps] = useBehavioralTrackingProps(remainingProps);
  const {
    "data-dt-name": dataDtName = "Overlay",
    ...remainingBehavioralTrackingProps
  } = behavioralTrackingProps;
  const [tooltipRef, setTooltipRef] = useState(null);
  const mergedTooltipRef = useMergeRefs([forwardedRef, setTooltipRef]);
  const fallbackPlacements = useMemo(() => {
    if (fallbackPlacementsProp) {
      return fallbackPlacementsProp;
    }
    switch (placement) {
      case "top":
        return ["bottom", "left", "right"];
      case "bottom":
        return ["top", "left", "right"];
      case "left":
        return ["right", "top", "bottom"];
      case "right":
        return ["left", "top", "bottom"];
    }
  }, [placement, fallbackPlacementsProp]);
  const { styles, attributes } = usePopper(triggerElement, tooltipRef, {
    placement,
    modifiers: [
      {
        name: "offset",
        options: {
          offset: [0, TOOLTIP_OFFSET]
        }
      },
      {
        name: "flip",
        options: {
          fallbackPlacements
        }
      },
      observeOverlaySize,
      observeAnimationDone,
      eventListeners
    ]
  });
  const stackingContext = useOverlayStacking(void 0, "overlay");
  const tooltipHovered = useRef(false);
  const handleMouseLeaveTooltip = useCallback(
    (e) => {
      tooltipHovered.current = false;
      onMouseLeave?.(e);
      onTooltipHovered?.(tooltipHovered.current);
    },
    [onMouseLeave, onTooltipHovered]
  );
  return /* @__PURE__ */ jsx(
    "div",
    {
      ref: mergedTooltipRef,
      style: {
        ...styles.popper,
        zIndex: stackingContext.style.zIndex,
        ...consumerStyle,
        ...assignInlineVars({ [tooltipOffsetVar]: `${TOOLTIP_OFFSET}px` })
      },
      className: clsx(tooltipElementCSS, consumerClassName),
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      role: "tooltip",
      onMouseEnter: () => {
        tooltipHovered.current = true;
        onTooltipHovered?.(tooltipHovered.current);
      },
      onMouseLeave: handleMouseLeaveTooltip,
      ...attributes.popper,
      ...remainingBehavioralTrackingProps,
      "data-dt-name": dataDtName,
      "data-dt-component": dataDtComponent,
      children: /* @__PURE__ */ jsx("div", { className: tooltipAnimationWrapperCSS, children: typeof children === "string" ? /* @__PURE__ */ jsx(Text, { textStyle: "small", as: "p", children }) : children })
    }
  );
});
TooltipElement.displayName = "TooltipElement";
export {
  TooltipElement
};
