// @dynatrace/strato-components/charts/timeseries/types/timeseries-grid-area-rules.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var timeseries_grid_area_rules_exports = {};
__export(timeseries_grid_area_rules_exports, {
  AREAS: () => AREAS
});
module.exports = __toCommonJS(timeseries_grid_area_rules_exports);
const AREAS = {
  annotations: "annotations",
  leftAxis: "left-axis",
  leftThreshold: "left-threshold",
  bottomAxis: "bottom-axis",
  rightAxis: "right-axis",
  rightThreshold: "right-threshold"
};
