// @dynatrace/strato-components/charts/honeycomb/utils/build-overlay-tooltip-data.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var build_overlay_tooltip_data_exports = {};
__export(build_overlay_tooltip_data_exports, {
  buildOverlayTooltipData: () => buildOverlayTooltipData
});
module.exports = __toCommonJS(build_overlay_tooltip_data_exports);
var import_safe_get = require("../../core/utils/safe-get.js");
const buildOverlayTooltipData = (tileIndex, honeycombData, tooltipMetadata) => {
  const tileData = honeycombData[tileIndex];
  const metadata = (0, import_safe_get.safeGet)(tooltipMetadata, tileData);
  return {
    geometry: "honeycomb",
    header1: [],
    metadata: {
      id: metadata.id,
      tileData,
      tooltipMetadata: metadata
    },
    series: [
      {
        name: metadata.name,
        color: metadata.color.Default,
        value: metadata.value.toString()
      }
    ]
  };
};
