// @dynatrace/strato-components/charts/honeycomb/utils/iterate-honeycomb-slots.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var iterate_honeycomb_slots_exports = {};
__export(iterate_honeycomb_slots_exports, {
  iterateHoneycombSlots: () => iterateHoneycombSlots
});
module.exports = __toCommonJS(iterate_honeycomb_slots_exports);
var import_react = require("react");
var import_Intent = require("../../../core/slots/Intent/Intent.js");
var import_is_slot = require("../../../core/utils/is-slot.js");
var import_ColorRule = require("../../core/slots/ColorRule.js");
var import_ChartToolbar = require("../../core/slots/toolbar/ChartToolbar.js");
var import_CompactToolbar = require("../../core/slots/toolbar/CompactToolbar.js");
var import_DownloadData = require("../../core/slots/toolbar/DownloadData.js");
var import_extract_behavior_tracking_props = require("../../core/utils/extract-behavior-tracking-props.js");
var import_constants = require("../constants.js");
var import_DownloadCSV = require("../slots/DownloadCSV.js");
var import_HoneycombLegend = require("../slots/HoneycombLegend.js");
var import_Toolbar = require("../slots/Toolbar.js");
var import_Tooltip = require("../slots/Tooltip.js");
const buildChartToolbarConfigSlot = (child, config) => {
  const { children, ...toolbar } = child.props;
  const childrenToolbar = iterateChartToolbarSlots(children);
  return {
    ...config,
    toolbar: {
      ...toolbar,
      ...childrenToolbar
    }
  };
};
const iterateHoneycombSlots = (children) => {
  let config = {
    legend: void 0,
    toolbar: void 0,
    colorRules: [],
    tooltip: void 0
  };
  let intents;
  import_react.Children.forEach(children, (child) => {
    if ((0, import_is_slot.isSlot)(child, import_HoneycombLegend.HoneycombLegend)) {
      config = {
        ...config,
        legend: {
          position: import_constants.DEFAULT_LEGEND_POSITION,
          ...child.props
        }
      };
    } else if ((0, import_is_slot.isSlot)(child, import_ChartToolbar.ChartToolbar)) {
      config = buildChartToolbarConfigSlot(child, config);
    } else if ((0, import_is_slot.isSlot)(child, import_Toolbar.Toolbar)) {
      config.toolbar = {
        ...config.toolbar,
        ...(0, import_CompactToolbar.sanitizeProps)(child.props, ["draggable", "hidden", "placement"]),
        ...(0, import_extract_behavior_tracking_props.extractBehaviorTrackingProps)(child.props)
      };
    } else if ((0, import_is_slot.isSlot)(child, import_DownloadCSV.HoneycombChartDownloadCSVSlot)) {
      config = {
        ...config,
        toolbar: { ...config.toolbar, downloadCSV: { enabled: true } }
      };
    } else if ((0, import_is_slot.isSlot)(child, import_Intent.Intent)) {
      intents = intents ? [...intents, child.props] : [child.props];
    } else if ((0, import_is_slot.isSlot)(child, import_ColorRule.ColorRule)) {
      (0, import_ColorRule.isValidColorRule)(child.props) && config.colorRules.push(child.props);
    } else if ((0, import_is_slot.isSlot)(child, import_Tooltip.Tooltip)) {
      config = {
        ...config,
        tooltip: {
          hidden: child.props?.hidden ?? false,
          template: child.props?.children
        }
      };
    }
  });
  if (intents) {
    config = {
      ...config,
      toolbar: {
        ...config.toolbar,
        intents,
        hidden: config.toolbar?.hidden ?? false
      }
    };
  }
  return config;
};
const iterateChartToolbarSlots = (children) => {
  let toolbar = {};
  import_react.Children.forEach(children, (childToolbar) => {
    if ((0, import_is_slot.isSlot)(childToolbar, import_DownloadData.DownloadData)) {
      toolbar = {
        ...toolbar,
        downloadCSV: { enabled: true }
      };
    }
  });
  return toolbar;
};
