// @dynatrace/strato-components/filters/filter-field/error-fallback/ErrorFallback.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ErrorFallback_exports = {};
__export(ErrorFallback_exports, {
  ErrorFallback: () => ErrorFallback
});
module.exports = __toCommonJS(ErrorFallback_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_intl = require("react-intl");
var import_borders = __toESM(require("@dynatrace/strato-design-tokens/borders"));
var import_strato_icons = require("@dynatrace/strato-icons");
var import_messages = require("./messages.js");
var import_Button = require("../../../buttons/button/Button.js");
var import_Container = require("../../../layouts/container/Container.js");
var import_Flex = require("../../../layouts/flex/Flex.js");
var import_Text = require("../../../typography/text/Text.js");
const ErrorFallback = ({ resetErrorBoundary }) => {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Container.Container,
    {
      padding: 2,
      paddingLeft: 4,
      color: "critical",
      style: {
        borderRadius: import_borders.default.Radius.Container.Subdued
      },
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
        import_Flex.Flex,
        {
          flexDirection: "row",
          justifyContent: "space-between",
          alignItems: "center",
          gap: 4,
          children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.CriticalIcon, {}),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_Text.Text,
              {
                textStyle: "base-emphasized",
                style: {
                  flex: 1,
                  overflow: "hidden",
                  whiteSpace: "nowrap",
                  textOverflow: "ellipsis"
                },
                children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_intl.FormattedMessage, { ...import_messages.errorFallbackMessages.errorTitleMessage })
              }
            ),
            /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
              import_Button.Button,
              {
                variant: "emphasized",
                color: "critical",
                onClick: resetErrorBoundary,
                size: "condensed",
                style: { borderRadius: import_borders.default.Radius.Field.Subdued },
                children: [
                  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ResetIcon, {}) }),
                  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_intl.FormattedMessage, { ...import_messages.errorFallbackMessages.recoverStateButtonText })
                ]
              }
            )
          ]
        }
      )
    }
  );
};
