// @dynatrace/strato-components/filters/filter-field/error-fallback/filter-value-obfuscation-util.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var filter_value_obfuscation_util_exports = {};
__export(filter_value_obfuscation_util_exports, {
  obfuscateFilterString: () => obfuscateFilterString
});
module.exports = __toCommonJS(filter_value_obfuscation_util_exports);
var import_string_tree_conversion = require("../utils/conversion/string-tree-conversion.js");
var import_tree_string_conversion = require("../utils/conversion/tree-string-conversion.js");
var import_escape_suggestion = require("../utils/escape-suggestion.js");
const TYPE_REPLACEMENT_MAP = {
  Key: "k",
  String: "s",
  JSONPath: "j",
  Variable: "v",
  Number: "n",
  BigInt: "b",
  Duration: "d",
  IPAddress: "i",
  Error: "e",
  UID: "u",
  SmartscapeId: "c",
  Timestamp: "t",
  // The following are not obfuscated
  LogicalOperator: null,
  ComparisonOperator: null,
  ExistsOperator: null,
  SearchOperator: null,
  InclusionOperator: null,
  Boolean: null
};
const EXTENDED_RESERVED_CHARACTERS = `.${import_escape_suggestion.RESERVED_CHARACTERS}`;
function getReplacementChar(node) {
  if (node.type in TYPE_REPLACEMENT_MAP) {
    return TYPE_REPLACEMENT_MAP[node.type];
  }
  return "U";
}
function obfuscateText(textValue, replacementChar) {
  if (textValue === void 0) {
    return "";
  }
  return textValue.split("").map((char) => {
    if (EXTENDED_RESERVED_CHARACTERS.includes(char)) {
      return char;
    }
    return replacementChar;
  }).join("");
}
function obfuscateFilterFieldLeafNode(node) {
  if (node === void 0) {
    return node;
  }
  if (node.type === "List") {
    return {
      ...node,
      value: node.value.map((value) => obfuscateFilterFieldLeafNode(value))
    };
  }
  const replacementChar = getReplacementChar(node);
  if (!replacementChar) {
    return node;
  }
  return {
    ...node,
    // Special handling for Duration type to avoid obfuscating reserved characters like 's', 'm', 'h', etc.
    textValue: node.type === "Duration" ? "" : obfuscateText(node.textValue, replacementChar),
    value: obfuscateText(String(node.value), replacementChar)
  };
}
function obfuscateStatement(node) {
  return {
    ...node,
    key: obfuscateFilterFieldLeafNode(node.key),
    comparisonOperator: node.comparisonOperator,
    value: obfuscateFilterFieldLeafNode(node.value)
  };
}
function obfuscateGroupNode(groupNode) {
  return {
    ...groupNode,
    children: groupNode.children.map((node) => {
      if (node.type === "Group") {
        return obfuscateGroupNode(node);
      }
      if (node.type === "Statement") {
        return obfuscateStatement(node);
      }
      return node;
    })
  };
}
function obfuscateFilterString(filterString, validatorMap, parserConfig) {
  const tree = (0, import_string_tree_conversion.convertStringToFilterFieldTree)(filterString, {
    validatorMap,
    parserConfig
  });
  if (tree.tree) {
    return (0, import_tree_string_conversion.convertFilterFieldTreeToString)(obfuscateGroupNode(tree.tree));
  }
  return "no filter tree";
}
