// @dynatrace/strato-components/filters/filter-field/error-fallback/messages.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var messages_exports = {};
__export(messages_exports, {
  errorFallbackMessages: () => errorFallbackMessages
});
module.exports = __toCommonJS(messages_exports);
var import_react_intl = require("react-intl");
const errorFallbackMessages = (0, import_react_intl.defineMessages)({
  errorTitleMessage: {
    defaultMessage: "An error occurred",
    id: "JYch5LSreqEKVk0q",
    description: "The message that is shown once the filter field reaches an invalid state when an uncaught exception was thrown"
  },
  recoverStateButtonText: {
    defaultMessage: "Reset",
    id: "IigYU5spaVAkF/1K",
    description: "Button text for trying to recover the filter field state from an invalid state"
  }
});
