// @dynatrace/strato-components/filters/timeframe-selector/components/presets/PresetItem.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var PresetItem_exports = {};
__export(PresetItem_exports, {
  PresetItem: () => PresetItem
});
module.exports = __toCommonJS(PresetItem_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var styles = __toESM(require("./Presets.sty.js"));
var import_useId = require("../../../../core/hooks/useId.js");
var import_useFocusRing = require("../../../../core/styles/useFocusRing.js");
var import_useTimeframeStore = require("../../../../core/time/timeframeStore/useTimeframeStore.js");
var import_merge_props = require("../../../../core/utils/merge-props.js");
var import_timeParser = require("../../../../forms/forms-core/time/timeParser.js");
var import_Flex = require("../../../../layouts/flex/Flex.js");
var import_Grid = require("../../../../layouts/grid/Grid.js");
var import_Tooltip = require("../../../../overlays/tooltip/Tooltip.js");
var import_Text = require("../../../../typography/text/Text.js");
var import_timeframeSelectorMessages = require("../../constants/timeframeSelectorMessages.js");
var import_usePresetItemContext = require("../../hooks/usePresetItemContext.js");
var import_useTimeframeSelectorContext = require("../../hooks/useTimeframeSelectorContext.js");
var import_useTooltipMessages = require("../../hooks/useTooltipMessages.js");
var import_presets_helpers = require("../../utils/presets-helpers.js");
var import_timeframe_selection = require("../../utils/timeframe-selection.js");
const PresetItem = (props) => {
  const {
    children,
    value,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    ...remainingProps
  } = props;
  const intl = (0, import_react_intl.useIntl)();
  const [openTooltip, setOpenTooltip] = (0, import_react.useState)(false);
  const { details } = (0, import_useTimeframeSelectorContext.useTimeframeSelectorContext)();
  const setValue = (0, import_useTimeframeStore.useTimeframeSelectorStore)(
    (state2) => state2.setTimeframeSelectorTimeframe
  );
  const min = (0, import_useTimeframeStore.useTimeframeSelectorStore)((state2) => state2.min?.value);
  const max = (0, import_useTimeframeStore.useTimeframeSelectorStore)((state2) => state2.max?.value);
  const precision = (0, import_useTimeframeStore.useTimeframeSelectorStore)((state2) => state2.precision);
  const { isRecent } = (0, import_usePresetItemContext.usePresetItemContext)();
  const isSelected = (0, import_presets_helpers.isPresetSelected)(value, details, isRecent, precision);
  const isDisabled = (0, import_timeframe_selection.isOutsideRange)(value, min, max);
  const tooltipId = (0, import_useId.useId)();
  const tooltipFocusProps = {
    onFocus: (0, import_react.useCallback)(
      () => !isDisabled && setOpenTooltip(true),
      [isDisabled]
    ),
    onBlur: (0, import_react.useCallback)(() => setOpenTooltip(false), [])
  };
  const { focusProps, focusClassName } = (0, import_useFocusRing.useFocusRing)({
    color: isSelected && !isDisabled ? "primary" : "neutral",
    disabled: isDisabled
  });
  let state = isSelected ? "selected" : "rest";
  if (isDisabled) {
    state = "disabled";
  }
  const {
    timeframeDisplayValue,
    formattedPresetItemValue,
    tooltipFormattedTimeValueStart,
    tooltipFormattedTimeValueEnd,
    isAbsoluteTimeframe,
    isInvalid
  } = (0, import_useTooltipMessages.useTooltipMessages)(value, children);
  const handleItemClick = (0, import_react.useCallback)(() => {
    if (isDisabled) {
      return;
    }
    const { from, to } = (0, import_timeParser.parseValue)(value);
    if (from && to) {
      const timeframe = {
        from: (0, import_timeParser.removePrecisionFromTimeValue)(from, precision) ?? from,
        to: (0, import_timeParser.removePrecisionFromTimeValue)(to, precision) ?? to
      };
      setValue(timeframe);
      (0, import_presets_helpers.saveRecentTimeframe)(timeframe);
    }
  }, [isDisabled, value, precision, setValue]);
  const handleItemKeyDown = (0, import_react.useCallback)(
    (event) => {
      if (isDisabled) {
        return;
      }
      if (event.code === "Enter" || event.code === "NumpadEnter" || event.code === "Space") {
        handleItemClick();
      }
    },
    [handleItemClick, isDisabled]
  );
  const liRef = (0, import_react.useRef)(null);
  (0, import_react.useEffect)(() => {
    if (isSelected) {
      liRef.current?.scrollIntoView({ block: "nearest" });
    }
  }, []);
  if (isInvalid) {
    return null;
  }
  const tooltipContent = isDisabled ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { textStyle: "small", children: intl.formatMessage(import_timeframeSelectorMessages.TIMEFRAME_SELECTOR_MESSAGES.disabledTooltip) }) : /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { flexDirection: "column", children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { gap: 0, flexDirection: "column", children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { textStyle: "base-emphasized", children: timeframeDisplayValue }),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { textStyle: "small", children: formattedPresetItemValue })
    ] }),
    !isAbsoluteTimeframe && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { gap: 0, flexDirection: "column", children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { textStyle: "small", children: tooltipFormattedTimeValueStart }),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { textStyle: "small", children: tooltipFormattedTimeValueEnd })
    ] })
  ] });
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "li",
    {
      ref: liRef,
      role: "option",
      "aria-selected": isSelected,
      "aria-disabled": isDisabled,
      className: (0, import_clsx.default)([
        focusClassName,
        styles.item({
          state
        }),
        consumerClassName
      ]),
      onClick: handleItemClick,
      onKeyDown: handleItemKeyDown,
      "data-testid": dataTestId,
      style: consumerStyle,
      ...(0, import_merge_props.mergeProps)(remainingProps, focusProps, tooltipFocusProps),
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_Tooltip.Tooltip,
        {
          id: tooltipId,
          text: tooltipContent,
          placement: "right",
          fallbackPlacements: [
            "right-end",
            "bottom-end",
            "bottom",
            "bottom-start",
            "left-end",
            "left",
            "left-start"
          ],
          open: openTooltip,
          onOpenChange: setOpenTooltip,
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "div",
            {
              className: styles.wrapper({ isSelected, isDisabled }),
              "aria-describedby": tooltipId,
              children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Grid.Grid, { className: styles.itemContainer, children: [
                /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { alignItems: "center", className: styles.displayValueWrapper, children: [
                  isRecent && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.HistoryIcon, { className: styles.recentIcon }),
                  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                    import_Text.Text,
                    {
                      textStyle: "base-emphasized",
                      as: "div",
                      className: styles.timeframeDisplayValue,
                      children: timeframeDisplayValue
                    }
                  )
                ] }),
                (!isRecent || !isAbsoluteTimeframe) && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  import_Text.Text,
                  {
                    as: import_Flex.Flex,
                    textStyle: "small",
                    className: (0, import_clsx.default)(
                      styles.mediaQuery,
                      styles.formattedPresetValue({
                        state
                      })
                    ),
                    children: formattedPresetItemValue
                  }
                )
              ] })
            }
          )
        }
      )
    }
  );
};
PresetItem.displayName = "TimeframeSelector.PresetItem";
