// @dynatrace/strato-components/filters/timeframe-selector/components/presets/Presets.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Presets_exports = {};
__export(Presets_exports, {
  Presets: () => Presets
});
module.exports = __toCommonJS(Presets_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_PresetItem = require("./PresetItem.js");
var styles = __toESM(require("./Presets.sty.js"));
var import_is_slot = require("../../../../core/utils/is-slot.js");
var import_logging = require("../../../../core/utils/logging.js");
var import_timeframe_presets = require("../../constants/timeframe-presets.js");
var import_timeframeSelectorMessages = require("../../constants/timeframeSelectorMessages.js");
function getItemSlots(children) {
  if (!children) {
    return null;
  }
  const validChildren = [];
  for (const child of import_react.Children.toArray(children)) {
    if ((0, import_is_slot.isSlot)(child, import_PresetItem.PresetItem)) {
      validChildren.push(child);
    } else {
      (0, import_logging.devLog)(
        "The used child is not supported by the TimeframeSelector.Presets."
      );
    }
  }
  return validChildren.length > 0 ? validChildren : null;
}
const Presets = (props) => {
  const {
    children,
    "aria-label": ariaLabel,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    ...remainingProps
  } = props;
  const intl = (0, import_react_intl.useIntl)();
  const presetsListLabel = ariaLabel ?? intl.formatMessage(import_timeframeSelectorMessages.TIMEFRAME_SELECTOR_MESSAGES.timeframePresets);
  const validChildren = getItemSlots(children);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "ul",
    {
      role: "listbox",
      "aria-label": presetsListLabel,
      className: (0, import_clsx.default)([styles.listContainer, consumerClassName]),
      style: consumerStyle,
      "data-testid": dataTestId,
      ...remainingProps,
      children: validChildren ?? import_timeframe_presets.TIMEFRAME_SELECTOR_PRESETS.map((item) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_PresetItem.PresetItem,
        {
          value: item,
          "data-dt-name": "Preset",
          "data-dt-properties": `valueFrom:${item.from},valueTo:${item.to}`
        },
        `${item.from} \u2192 ${item.to}`
      ))
    }
  );
};
Presets["displayName"] = "TimeframeSelector.Presets";
