// @dynatrace/strato-components/filters/timeframe-selector/components/presets/Presets.sty.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Presets_css_exports = {};
__export(Presets_css_exports, {
  displayValueWrapper: () => displayValueWrapper,
  formattedPresetValue: () => formattedPresetValue,
  indicatorIcon: () => indicatorIcon,
  item: () => item,
  itemContainer: () => itemContainer,
  listContainer: () => listContainer,
  mediaQuery: () => mediaQuery,
  recentHeading: () => recentHeading,
  recentIcon: () => recentIcon,
  timeframeDisplayValue: () => timeframeDisplayValue,
  wrapper: () => wrapper
});
module.exports = __toCommonJS(Presets_css_exports);
var import_Presets_css_ts_vanilla = require("./Presets.css");
var import_createRuntimeFn = require("@vanilla-extract/recipes/createRuntimeFn");
var displayValueWrapper = "_14rlfn36-3-6-0";
var formattedPresetValue = (0, import_createRuntimeFn.createRuntimeFn)({ defaultClassName: "_14rlfn3e-3-6-0", variantClassNames: { state: { rest: "_14rlfn3f-3-6-0", selected: "_14rlfn3g-3-6-0", disabled: "_14rlfn3h-3-6-0" } }, defaultVariants: { state: "rest" }, compoundVariants: [] });
var indicatorIcon = (0, import_createRuntimeFn.createRuntimeFn)({ defaultClassName: "_14rlfn3i-3-6-0", variantClassNames: { isDisabled: { true: "_14rlfn3j-3-6-0", false: "_14rlfn3k-3-6-0" } }, defaultVariants: {}, compoundVariants: [] });
var item = (0, import_createRuntimeFn.createRuntimeFn)({ defaultClassName: "_14rlfn31-3-6-0", variantClassNames: { state: { rest: "_14rlfn32-3-6-0", selected: "_14rlfn33-3-6-0", disabled: "_14rlfn34-3-6-0" }, defaultVariants: { state: "rest" } }, defaultVariants: {}, compoundVariants: [] });
var itemContainer = "_14rlfn30-3-6-0";
var listContainer = "_14rlfn3l-3-6-0";
var mediaQuery = "_14rlfn3d-3-6-0";
var recentHeading = "_14rlfn3m-3-6-0";
var recentIcon = "_14rlfn3n-3-6-0";
var timeframeDisplayValue = "_14rlfn37-3-6-0";
var wrapper = (0, import_createRuntimeFn.createRuntimeFn)({ defaultClassName: "_14rlfn38-3-6-0", variantClassNames: { isSelected: { true: "_14rlfn39-3-6-0", false: "_14rlfn3a-3-6-0" }, isDisabled: { true: "_14rlfn3b-3-6-0", false: "_14rlfn3c-3-6-0" } }, defaultVariants: {}, compoundVariants: [] });
