// @dynatrace/strato-components/filters/timeframe-selector/components/presets/RecentTimeframes.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var RecentTimeframes_exports = {};
__export(RecentTimeframes_exports, {
  RecentTimeframes: () => RecentTimeframes
});
module.exports = __toCommonJS(RecentTimeframes_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_query = require("@tanstack/react-query");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_PresetItem = require("./PresetItem.js");
var styles = __toESM(require("./Presets.sty.js"));
var import_Divider = require("../../../../layouts/divider/Divider.js");
var import_Text = require("../../../../typography/text/Text.js");
var import_timeframeSelectorMessages = require("../../constants/timeframeSelectorMessages.js");
var import_PresetItemContext = require("../../contexts/PresetItemContext.js");
var import_presets_helpers = require("../../utils/presets-helpers.js");
var import_TimeframePresetsClient = require("../../utils/TimeframePresetsClient.js");
const RecentTimeframes = () => {
  const intl = (0, import_react_intl.useIntl)();
  const { data, isLoading } = (0, import_react_query.useQuery)(
    {
      queryKey: import_TimeframePresetsClient.TIMEFRAME_PRESETS_QUERY_KEY,
      queryFn: import_presets_helpers.getTimeframePresets,
      notifyOnChangeProps: ["dataUpdatedAt"]
    },
    import_TimeframePresetsClient.timeframeQueryClient
  );
  const presetItemContextValue = (0, import_react.useMemo)(
    () => ({ isRecent: true }),
    []
  );
  if (!data?.recentPresets?.length || isLoading) {
    return null;
  }
  const ariaLabel = intl.formatMessage(
    import_timeframeSelectorMessages.TIMEFRAME_SELECTOR_MESSAGES.recentTimeframesList
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Divider.Divider, { flexItem: false }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { textStyle: "small", className: styles.recentHeading, as: "div", children: intl.formatMessage(
        import_timeframeSelectorMessages.TIMEFRAME_SELECTOR_MESSAGES.recentTimeframesHeading
      ) }),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        "ul",
        {
          role: "listbox",
          className: styles.listContainer,
          "aria-label": ariaLabel,
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_PresetItemContext.PresetItemContext.Provider, { value: presetItemContextValue, children: data.recentPresets.map(({ timeframe }, index) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_PresetItem.PresetItem,
            {
              value: timeframe,
              "data-dt-name": "Recent",
              "data-dt-properties": `index:${index},valueFrom:${timeframe.from},valueTo:${timeframe.to}`
            },
            `${timeframe.from} \u2192 ${timeframe.to}`
          )) })
        }
      )
    ] })
  ] });
};
RecentTimeframes.displayName = "RecentTimeframes";
