// @dynatrace/strato-components/esm/charts/tree-map/components/TreeMapInternal.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { Treemap, hierarchy, treemapSquarify } from "@visx/hierarchy";
import {
  forwardRef,
  useMemo
} from "react";
import { TreeMapCanvas } from "./TreeMapCanvas.js";
import { TreeMapCluster } from "./TreeMapCluster.js";
import { Canvas } from "../../core/components/canvas/Canvas.js";
import { CanvasContext } from "../../core/components/canvas/CanvasContext.js";
import { useGraphSize } from "../../core/components/chart-layout/hooks/useGraphSize.js";
import { TREEMAP_CHART_ARIA_LABEL } from "../constants.js";
import { useLabelsDisplay } from "../context/LabelsDisplayContext.js";
import { useTreeMapAggregationData } from "../context/TreeMapAggregationDataContext.js";
import { useTreeMapHighlightingMap } from "../context/TreeMapHighlightingMapContext.js";
import { useHighlightedState } from "../store/selectors.js";
import {
  nodePadding,
  nodePaddingOuter,
  nodePaddingTop
} from "../utils/node-padding.js";
import { withParentLabel } from "../utils/with-parent-label.js";
const TreeMapInternal = forwardRef(
  (props, forwardedRef) => {
    const { children } = props;
    const { width, height } = useGraphSize();
    const roundedWidth = Math.floor(width);
    const roundedHeight = Math.floor(height);
    const labelsDisplay = useLabelsDisplay();
    const { dataForRendering } = useTreeMapAggregationData();
    const highlightingMap = useTreeMapHighlightingMap();
    const hasParentLabels = withParentLabel(labelsDisplay);
    const root = useMemo(() => hierarchy(dataForRendering), [dataForRendering]);
    const highlightedItem = useHighlightedState();
    return /* @__PURE__ */ jsxs("div", { ref: forwardedRef, "aria-label": TREEMAP_CHART_ARIA_LABEL, role: "img", children: [
      children,
      /* @__PURE__ */ jsx(
        CanvasContext.Provider,
        {
          value: {
            shapeRefs: /* @__PURE__ */ new Map()
          },
          children: /* @__PURE__ */ jsx(Canvas, { width: roundedWidth, height: roundedHeight, children: /* @__PURE__ */ jsx(
            Treemap,
            {
              root,
              size: [roundedWidth, roundedHeight],
              tile: treemapSquarify,
              padding: (node) => nodePadding(node, hasParentLabels),
              paddingTop: hasParentLabels ? (node) => nodePaddingTop(node) : void 0,
              paddingOuter: hasParentLabels ? (node) => nodePaddingOuter(node) : void 0,
              round: true,
              children: (rootNode) => /* @__PURE__ */ jsxs(Fragment, { children: [
                hasParentLabels && /* @__PURE__ */ jsx(
                  TreeMapCluster,
                  {
                    treeMapData: rootNode,
                    highlightedItem,
                    highlightingMap
                  }
                ),
                /* @__PURE__ */ jsx(
                  TreeMapCanvas,
                  {
                    treeMapData: rootNode,
                    highlightedItem,
                    highlightingMap
                  }
                )
              ] })
            }
          ) })
        }
      )
    ] });
  }
);
export {
  TreeMapInternal
};
