// @dynatrace/strato-components/esm/charts/tree-map/components/TreeMapChartLayout.js
import { jsx, jsxs } from "react/jsx-runtime";
import { forwardRef } from "react";
import { TreeMapInternal } from "./TreeMapInternal.js";
import { TreeMapLegendRenderer } from "./TreeMapLegend/TreeMapLegendRenderer.js";
import { TreeMapOverlay } from "./TreeMapOverlay/TreeMapOverlay.js";
import { useBehavioralTrackingProps } from "../../../core/hooks/useBehavioralTrackingProps.js";
import { ChartLayout } from "../../core/components/chart-layout/ChartLayout.js";
import { OverlayProvider } from "../../core/components/overlay-tooltip/contexts/OverlayContext.js";
import { TooltipProvider } from "../../core/components/overlay-tooltip/contexts/TooltipContext.js";
import { useChartLegend } from "../context/ChartLegendContext.js";
import { TreeMapAggregationDataProvider } from "../providers/TreeMapAggregationDataProvider.js";
import { TreeMapHighlightingMapProvider } from "../providers/TreeMapHighlightingMapProvider.js";
import { TreeMapInternalDataProvider } from "../providers/TreeMapInternalDataProvider.js";
const TreeMapChartLayout = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const {
    chartLayoutRef,
    width,
    height,
    emptyState,
    errorState,
    isEmpty,
    loading,
    className,
    style,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "data-testid": dataTestId,
    "data-dt-component": dataDtComponent
  } = props;
  const [behavioralTrackingProps] = useBehavioralTrackingProps(props);
  const {
    hidden: legendHidden,
    position: legendPosition,
    // eslint-disable-next-line @typescript-eslint/no-deprecated
    ratio,
    initialRatio,
    onRatioChange
  } = useChartLegend();
  const legendRatio = initialRatio ?? ratio;
  return /* @__PURE__ */ jsxs(
    ChartLayout,
    {
      ...behavioralTrackingProps,
      ref: chartLayoutRef,
      chartWidth: width,
      chartHeight: height,
      emptyState,
      errorState,
      showLoader: loading,
      isEmpty,
      className,
      style,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "data-testid": dataTestId,
      "data-dt-component": dataDtComponent,
      children: [
        /* @__PURE__ */ jsx(ChartLayout.Graph, { children: /* @__PURE__ */ jsx(TreeMapInternalDataProvider, { children: /* @__PURE__ */ jsx(TreeMapAggregationDataProvider, { children: /* @__PURE__ */ jsx(TreeMapHighlightingMapProvider, { children: /* @__PURE__ */ jsx(OverlayProvider, { children: /* @__PURE__ */ jsx(TooltipProvider, { children: /* @__PURE__ */ jsxs("div", { style: { position: "relative" }, children: [
          /* @__PURE__ */ jsx(TreeMapInternal, { ref: forwardedRef }),
          loading ? null : /* @__PURE__ */ jsx(TreeMapOverlay, {})
        ] }) }) }) }) }) }) }),
        !legendHidden && /* @__PURE__ */ jsx(
          ChartLayout.Legend,
          {
            position: legendPosition,
            initialRatio: legendRatio,
            onResize: onRatioChange,
            children: /* @__PURE__ */ jsx(TreeMapLegendRenderer, {})
          }
        )
      ]
    }
  );
});
export {
  TreeMapChartLayout
};
