// @dynatrace/strato-components/esm/charts/pie/types/default.js
import Spacings from "@dynatrace/strato-design-tokens/spacings";
import { DEFAULT_TRUNCATION_MODE, defaultRatio } from "../../core/constants.js";
const DEFAULT_CONTAINER_PADDING = Spacings.Size6;
const DEFAULT_SIZE = 300;
const DONUT_INTERNAL_RADIUS = 0.75;
const PIE_INTERNAL_RADIUS = 1;
const LABEL_FONT_SIZE = 12;
const SHOW_LABEL_RELATIVE_VALUE_THRESHOLD = 20;
const SHOW_LABEL_SLICES_THRESHOLD = 2;
const LABEL_FONT_SIZE_MULTIPLIER = 1.5;
const DEFAULT_LEGEND = {
  position: "auto",
  hidden: false,
  ratio: defaultRatio
};
const DEFAULT_TOOLTIP_MARGIN = 40;
const defaultConfig = {
  height: DEFAULT_SIZE,
  colorPalette: "categorical",
  labels: { hidden: false, valueType: "relative" },
  grouping: { threshold: { type: "relative", value: 2 } },
  legend: DEFAULT_LEGEND,
  formatter: () => "",
  truncationMode: DEFAULT_TRUNCATION_MODE
};
const zIndicesPieAreas = {
  toolbar: 2,
  overlay: 1
};
const THRESHOLD_TYPE = {
  ABSOLUTE: "absolute",
  RELATIVE: "relative",
  NUMBER_OF_SLICES: "number-of-slices"
};
const PIE_CATEGORICAL_LEGEND_SIZES = {
  minWidth: 0.1,
  minHeight: 0.1,
  maxWidth: 0.8,
  maxHeight: 0.7
};
const LABEL_TEXT_POSITION = {
  END: "end",
  START: "start"
};
export {
  DEFAULT_CONTAINER_PADDING,
  DEFAULT_LEGEND,
  DEFAULT_SIZE,
  DEFAULT_TOOLTIP_MARGIN,
  DONUT_INTERNAL_RADIUS,
  LABEL_FONT_SIZE,
  LABEL_FONT_SIZE_MULTIPLIER,
  LABEL_TEXT_POSITION,
  PIE_CATEGORICAL_LEGEND_SIZES,
  PIE_INTERNAL_RADIUS,
  SHOW_LABEL_RELATIVE_VALUE_THRESHOLD,
  SHOW_LABEL_SLICES_THRESHOLD,
  THRESHOLD_TYPE,
  defaultConfig,
  zIndicesPieAreas
};
