// @dynatrace/strato-components/typography/index.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var typography_exports = {};
__export(typography_exports, {
  Blockquote: () => import_Blockquote.Blockquote,
  Code: () => import_Code.Code,
  Emphasis: () => import_Emphasis.Emphasis,
  ExternalLink: () => import_ExternalLink.ExternalLink,
  Heading: () => import_Heading.Heading,
  Highlight: () => import_Highlight.Highlight,
  Link: () => import_Link.Link,
  List: () => import_List.List,
  Paragraph: () => import_Paragraph.Paragraph,
  Strikethrough: () => import_Strikethrough.Strikethrough,
  Strong: () => import_Strong.Strong,
  Text: () => import_Text.Text,
  TextEllipsis: () => import_TextEllipsis.TextEllipsis,
  _ELLIPSIS_APPROXIMATION_THRESHOLD: () => import_utils.ELLIPSIS_APPROXIMATION_THRESHOLD,
  _ELLIPSIS_CHARACTER: () => import_utils.ELLIPSIS_CHARACTER,
  _calculateEllipsisInfo: () => import_utils.calculateEllipsisInfo,
  _centerEllipsizeText: () => import_utils.centerEllipsizeText,
  _getTextHeight: () => import_utils.getTextHeight,
  _getTextWidth: () => import_utils.getTextWidth
});
module.exports = __toCommonJS(typography_exports);
var import_Blockquote = require("./block-quote/Blockquote.js");
var import_Code = require("./code/Code.js");
var import_Emphasis = require("./emphasis/Emphasis.js");
var import_ExternalLink = require("./external-link/ExternalLink.js");
var import_Heading = require("./heading/Heading.js");
var import_Highlight = require("./highlight/Highlight.js");
var import_Link = require("./link/Link.js");
var import_List = require("./list/List.js");
var import_Paragraph = require("./paragraph/Paragraph.js");
var import_Strikethrough = require("./strikethrough/Strikethrough.js");
var import_Strong = require("./strong/Strong.js");
var import_Text = require("./text/Text.js");
var import_TextEllipsis = require("./text-ellipsis/TextEllipsis.js");
var import_utils = require("./utils.js");
