// @dynatrace/strato-components/typography/utils.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var utils_exports = {};
__export(utils_exports, {
  ELLIPSIS_APPROXIMATION_THRESHOLD: () => ELLIPSIS_APPROXIMATION_THRESHOLD,
  ELLIPSIS_CHARACTER: () => ELLIPSIS_CHARACTER,
  calculateEllipsisInfo: () => calculateEllipsisInfo,
  centerEllipsizeText: () => centerEllipsizeText,
  getTextHeight: () => getTextHeight,
  getTextWidth: () => getTextWidth
});
module.exports = __toCommonJS(utils_exports);
var import_user_preferences = require("@dynatrace-sdk/user-preferences");
let localizedSegmenter;
function getLocalizedSegmenter() {
  if (localizedSegmenter) {
    return localizedSegmenter;
  }
  const segmenter = new Intl.Segmenter((0, import_user_preferences.getLanguage)(), {
    granularity: "grapheme"
  });
  localizedSegmenter = segmenter;
  return localizedSegmenter;
}
const ELLIPSIS_CHARACTER = "\u2026";
const ELLIPSIS_APPROXIMATION_THRESHOLD = 15;
let inMemContext2d;
function getTextWidth(text, textFont) {
  const context = getCanvasContext2d();
  context.font = getFontString(textFont);
  const textMetrics = context.measureText(text);
  if (textMetrics.width < 0) {
    return textMetrics.width;
  }
  const actualTextWidth = Math.abs(textMetrics.actualBoundingBoxLeft) + Math.abs(textMetrics.actualBoundingBoxRight);
  return Math.ceil(actualTextWidth);
}
function getTextHeight(text, textFont) {
  const context = getCanvasContext2d();
  context.font = getFontString(textFont);
  const textMetrics = context.measureText(text);
  const actualTextHeight = Math.abs(textMetrics.actualBoundingBoxAscent) + Math.abs(textMetrics.actualBoundingBoxDescent);
  return Math.ceil(actualTextHeight);
}
function centerEllipsizeText(originalText, textFont, maxWidthInPx) {
  const ellipsisInfo = calculateEllipsisInfo(
    originalText,
    textFont,
    maxWidthInPx
  );
  if (ellipsisInfo === null) {
    return originalText;
  }
  const { ellipsisWidth, avgLetterWidth } = ellipsisInfo;
  const availableTextWidth = maxWidthInPx - ellipsisWidth;
  if (availableTextWidth <= ellipsisWidth) {
    return ELLIPSIS_CHARACTER;
  }
  const initialPrefixSuffixLength = Math.floor(availableTextWidth / avgLetterWidth / 2) + 1;
  let prefix;
  let suffix;
  let ellipsizedText;
  let approximatedTextWidth;
  let attempt = 1;
  let prefixLength = initialPrefixSuffixLength;
  let suffixLength = initialPrefixSuffixLength;
  const segmenter = getLocalizedSegmenter();
  const originalSegmentedText = Array.from(segmenter.segment(originalText)).map(
    (segment) => segment.segment
  );
  do {
    const originalSegmentedTextCopy = [...originalSegmentedText];
    prefixLength -= attempt % 2 === 1 ? 1 : 0;
    suffixLength -= attempt % 2 === 0 ? 1 : 0;
    prefix = originalSegmentedTextCopy.slice(0, prefixLength).join("");
    suffix = originalSegmentedTextCopy.slice(suffixLength * -1).join("");
    ellipsizedText = `${prefix}${ELLIPSIS_CHARACTER}${suffix}`;
    approximatedTextWidth = getTextWidth(ellipsizedText, textFont);
    attempt++;
  } while (attempt <= ELLIPSIS_APPROXIMATION_THRESHOLD && approximatedTextWidth >= maxWidthInPx);
  return ellipsizedText;
}
function getFontString(textFont) {
  const { fontStyle, fontWeight, fontSize, fontFamily } = textFont;
  return [
    fontStyle === "normal" ? "" : fontStyle,
    fontWeight === "400" ? "" : fontWeight,
    fontSize,
    fontFamily
  ].filter((value) => value !== void 0 && value !== "").map(
    (value) => (
      // we need to resolve CSS variables manually because they are not supported by the canvas API
      value?.trim().startsWith("var(") ? value.replace(/^\s*var\(--[0-9A-Za-z-]+,\s*([^)]+)\)\s*$/, "$1") : value
    )
  ).join(" ");
}
function calculateEllipsisInfo(originalText, textFont, maxWidthInPx) {
  const segmenter = getLocalizedSegmenter();
  const originalTextLength = Array.from(segmenter.segment(originalText)).length;
  if (originalTextLength <= 1) {
    return null;
  }
  const textWidth = getTextWidth(originalText, textFont);
  const avgLetterWidth = textWidth / originalTextLength;
  return textWidth > maxWidthInPx ? {
    ellipsisWidth: getTextWidth(ELLIPSIS_CHARACTER, textFont),
    avgLetterWidth
  } : null;
}
function getCanvasContext2d() {
  if (!inMemContext2d) {
    const canvas = document.createElement("canvas");
    canvas.hidden = true;
    const context = canvas.getContext("2d", { alpha: false });
    if (context !== null) {
      context.imageSmoothingEnabled = false;
      context.imageSmoothingQuality = "low";
      inMemContext2d = context;
    } else {
      throw new Error("Failed to create a Canvas 2D context");
    }
  }
  return inMemContext2d;
}
