// @dynatrace/strato-components/charts/core/components/overlay-tooltip/components/series/SeriesItem.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SeriesItem_exports = {};
__export(SeriesItem_exports, {
  SeriesItem: () => SeriesItem
});
module.exports = __toCommonJS(SeriesItem_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_use_series_item_display_state = require("./hooks/use-series-item-display-state.js");
var import_SeriesActionsHeader = require("./SeriesActionsHeader.js");
var import_SeriesItemContent = require("./SeriesItemContent.js");
var import_useId = require("../../../../../../core/hooks/useId.js");
var import_copy_to_clipboard_messages = require("../../../copy-to-clipboard/copy-to-clipboard-messages.js");
var import_ChartActionsMenuAsync = require("../../../series-action/ChartActionsMenuAsync.js");
var import_constants = require("../../constants.js");
var import_ExternalConfigContext = require("../../contexts/ExternalConfigContext.js");
var import_useDefaultLabels = require("../../hooks/useDefaultLabels.js");
var import_selectors = require("../../selectors.js");
var import_tooltip_css = require("../../tooltip.sty.js");
var import_useTooltipReducer = require("../../useTooltipReducer.js");
var import_get_copy_action = require("../../utils/get-copy-action.js");
const ACTIVE_SERIES_HOVER_DELAY = 300;
const CHEVRON_SLOT_WIDTH = 20;
const SeriesItem = (props) => {
  const intl = (0, import_react_intl.useIntl)();
  const { pinned } = (0, import_selectors.useTooltipSnapshot)();
  const labelFor = (0, import_useDefaultLabels.useDefaultLabels)();
  const { actions, ...restProps } = props;
  const { symbolPosition = "left", defaultActionsConfig } = (0, import_ExternalConfigContext.useExternalConfig)();
  const hasActions = actions !== void 0 || defaultActionsConfig.size > 0;
  const dispatch = (0, import_useTooltipReducer.useTooltipReducer)();
  const rowId = (0, import_useId.useId)();
  const isOpen = (0, import_selectors.useSelector)(({ activeSeries: activeSeries2 }) => activeSeries2 === rowId);
  const { activeSeries } = (0, import_selectors.useTooltipSnapshot)();
  const menuDelayRef = (0, import_react.useRef)(0);
  const asyncActions = actions ? async () => actions() : void 0;
  const triggerLeftTooltip = (event) => {
    event.stopPropagation();
    dispatch({ type: "TOOLTIP_MOUSE_LEAVE" })();
  };
  const handleClick = (event) => {
    event.stopPropagation();
    menuDelayRef.current = 0;
    dispatch({
      type: "SET_ACTIVE_SERIES",
      payload: { activeSeries: rowId }
    })();
  };
  const handlePointerMove = () => {
    menuDelayRef.current = ACTIVE_SERIES_HOVER_DELAY;
    dispatch({
      type: "SET_ACTIVE_SERIES",
      payload: { activeSeries: rowId }
    })();
  };
  const defaultActions = !pinned ? null : [...defaultActionsConfig.entries()].map(([accessor, config]) => {
    const message = intl.formatMessage(
      config.messageLabel ?? import_copy_to_clipboard_messages.COPY_TO_CLIPBOARD_MESSAGES.copyField,
      {
        field: labelFor?.(
          config.field,
          config.customLabel
        )?.toLocaleLowerCase()
      }
    );
    return (0, import_get_copy_action.getDefaultAction)(props, accessor, message);
  });
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "button",
    {
      className: (0, import_clsx.default)(
        activeSeries === rowId && import_tooltip_css.seriesItemActive,
        import_tooltip_css.seriesItemWrapper
      ),
      onClick: handleClick,
      onPointerMove: handlePointerMove,
      onKeyDown: import_lodash_es.noop,
      "data-testid": "menu-trigger",
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_ChartActionsMenuAsync.ChartActionsMenuAsync,
        {
          hasActions,
          defaultActions,
          actions: asyncActions,
          isOpen,
          header: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_SeriesActionsHeader.SeriesActionsHeader, { text: props.name }),
          onActionSelected: triggerLeftTooltip,
          delay: menuDelayRef.current,
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            WithSymbol,
            {
              symbolPosition,
              hasActions,
              ...restProps
            }
          )
        }
      )
    }
  );
};
const WithSymbol = (0, import_react.forwardRef)(
  (props, ref) => {
    const containerWidth = import_tooltip_css.TOOLTIP_MAX_WIDTH - import_constants.TOOLTIP_CONTENT_PADDING;
    const {
      color,
      multiLine = false,
      name,
      description,
      shape,
      iconStyle,
      icon,
      symbol
    } = props;
    const { symbolPosition, hasActions } = props;
    const seriesContentMaxWidth = hasActions ? containerWidth - CHEVRON_SLOT_WIDTH : containerWidth;
    const {
      nameText,
      descriptionText,
      fullDescriptionText,
      isDescriptionTruncated,
      descriptionShouldUseMultiLineCss,
      descriptionMultiLineMode,
      displayValueText,
      displayRelativeValueText,
      fullValueText,
      fullRelativeValueText,
      nameTextStyle,
      nameStyle
    } = (0, import_use_series_item_display_state.useSeriesItemDisplayState)({
      ...props,
      maxWidth: seriesContentMaxWidth
    });
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { ref, className: import_tooltip_css.seriesItemWithChevron, children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_SeriesItemContent.SeriesItemContent,
        {
          color,
          name,
          description,
          nameText,
          descriptionText,
          fullDescriptionText,
          isDescriptionTruncated,
          descriptionShouldUseMultiLineCss,
          descriptionMultiLineMode,
          displayValueText,
          displayRelativeValueText,
          fullValueText,
          fullRelativeValueText,
          nameTextStyle,
          nameStyle,
          multiLine,
          symbolPosition,
          shape,
          symbol,
          iconStyle,
          icon,
          hasChevron: hasActions,
          maxWidth: seriesContentMaxWidth
        }
      ),
      hasActions && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        "span",
        {
          className: (0, import_clsx.default)(
            import_tooltip_css.chevron,
            symbolPosition === "right" && import_tooltip_css.chevronRightAligned
          ),
          "aria-hidden": true,
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ChevronRightSmallIcon, { size: 16 })
        }
      )
    ] });
  }
);
