// @dynatrace/strato-components/charts/core/components/overlay-tooltip/components/series/SeriesItemContent.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SeriesItemContent_exports = {};
__export(SeriesItemContent_exports, {
  SeriesItemContent: () => SeriesItemContent
});
module.exports = __toCommonJS(SeriesItemContent_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_dynamic = require("@vanilla-extract/dynamic");
var import_clsx = __toESM(require("clsx"));
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_SeriesSymbol = require("./SeriesSymbol.js");
var import_Tooltip = require("../../../../../../overlays/tooltip/Tooltip.js");
var import_constants = require("../../constants.js");
var import_tooltip_css = require("../../tooltip.sty.js");
const SeriesItemContent = (0, import_react.forwardRef)((props, ref) => {
  const {
    color,
    nameText,
    descriptionText,
    fullDescriptionText,
    isDescriptionTruncated,
    descriptionShouldUseMultiLineCss,
    descriptionMultiLineMode = "clamp",
    displayValueText,
    displayRelativeValueText,
    fullValueText,
    fullRelativeValueText,
    nameTextStyle,
    nameStyle,
    multiLine = false,
    symbolPosition,
    shape,
    symbol,
    icon,
    iconStyle,
    maxWidth,
    hasChevron = false
  } = props;
  const isRightPosition = symbolPosition === "right";
  const hasSymbol = color !== void 0 && !(0, import_lodash_es.isEmpty)(color) || icon !== void 0 || symbol !== void 0;
  const renderNameContent = () => {
    if (Array.isArray(nameText) && multiLine) {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("ul", { className: import_tooltip_css.multiLineName, children: nameText.map((line) => {
        return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          "li",
          {
            className: nameTextStyle,
            style: { ...nameStyle, padding: 0 },
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { className: import_tooltip_css.seriesNameText, children: line })
          },
          `${line}`
        );
      }) });
    }
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { className: nameTextStyle, style: nameStyle, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { className: import_tooltip_css.seriesNameText, children: nameText }) });
  };
  const renderSymbol = () => {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { className: (0, import_tooltip_css.symbolWrapper)({ isRight: isRightPosition }), children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_SeriesSymbol.SeriesSymbol,
      {
        color,
        shape,
        symbol,
        icon,
        iconStyle
      }
    ) });
  };
  const getDescriptionClassName = () => {
    if (!descriptionShouldUseMultiLineCss) {
      return import_tooltip_css.seriesDescriptionText;
    }
    return descriptionMultiLineMode === "wrap" ? import_tooltip_css.seriesDescriptionTextMultiLineWrap : import_tooltip_css.seriesDescriptionTextMultiLine;
  };
  const renderValue = () => {
    const isValueTruncated = displayValueText !== fullValueText;
    const isRelativeValueTruncated = displayRelativeValueText !== fullRelativeValueText;
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Tooltip.Tooltip, { text: fullValueText, disabled: !isValueTruncated, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { className: import_tooltip_css.seriesItemText, children: displayValueText }) }),
      displayRelativeValueText && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_Tooltip.Tooltip,
        {
          text: fullRelativeValueText,
          disabled: !isRelativeValueTruncated,
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { className: (0, import_clsx.default)(import_tooltip_css.relativeValueStyle, import_tooltip_css.seriesItemText), children: displayRelativeValueText })
        }
      )
    ] });
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    "div",
    {
      ref,
      className: (0, import_tooltip_css.seriesItem)({ symbolPosition, hasChevron }),
      style: (0, import_dynamic.assignInlineVars)({
        [import_tooltip_css.seriesItemMaxWidthVar]: `${maxWidth}px`,
        [import_tooltip_css.spaceSavedForSymbolVar]: hasSymbol ? "24px" : "6px"
      }),
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { style: { gridArea: "symbol" }, children: hasSymbol ? renderSymbol() : null }),
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
          "div",
          {
            className: import_tooltip_css.seriesNameCell,
            style: {
              gridArea: "name",
              marginRight: `${import_constants.GAP_BETWEEN_COLUMNS}px`,
              flexDirection: "column"
            },
            children: [
              renderNameContent(),
              descriptionText !== void 0 && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_Tooltip.Tooltip,
                {
                  text: fullDescriptionText ?? "",
                  disabled: !isDescriptionTruncated,
                  children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                    "span",
                    {
                      className: getDescriptionClassName(),
                      style: descriptionShouldUseMultiLineCss ? nameStyle : void 0,
                      children: descriptionText
                    }
                  )
                }
              )
            ]
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: import_tooltip_css.seriesValueCell, style: { gridArea: "value" }, children: renderValue() })
      ]
    }
  );
});
