// @dynatrace/strato-components/charts/core/components/overlay-tooltip/components/series/Series.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Series_exports = {};
__export(Series_exports, {
  Series: () => Series
});
module.exports = __toCommonJS(Series_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_virtual = require("@tanstack/react-virtual");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_constants = require("../../constants.js");
var import_ExternalConfigContext = require("../../contexts/ExternalConfigContext.js");
var import_use_scrollbar_detection = require("../../hooks/use-scrollbar-detection.js");
var import_selectors = require("../../selectors.js");
var import_tooltip_css = require("../../tooltip.sty.js");
var import_estimate_series_item_size = require("../../utils/estimate-series-item-size.js");
var import_width_calculation = require("../../utils/width-calculation.js");
const Series = (props) => {
  const { data, seriesItemTemplate, activeItem = 0 } = props;
  const config = (0, import_ExternalConfigContext.useExternalConfig)();
  if (config.variant !== "shared" || data.length === 1) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { className: import_tooltip_css.seriesGridContainer, children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react.Fragment, { children: config.seriesHeader }, "seriesHeader"),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react.Fragment, { children: seriesItemTemplate(data[0]) }, "series")
    ] });
  }
  const maxContentWidthWithinFirstTenValues = Math.max(
    ...data.slice(0, 10).map((item) => {
      let nameWidth = 0;
      if (Array.isArray(item.name)) {
        nameWidth = (0, import_width_calculation.calculateTotalWidth400)(...item.name);
      } else {
        nameWidth = (0, import_width_calculation.calculateTotalWidth400)(item.name);
      }
      let valueWidth = 0;
      if (item.value !== void 0 && item.value !== null) {
        const valueText = String(item.value);
        const relativeValueText = item.relativeValue ? String(item.relativeValue) : "";
        valueWidth = (0, import_width_calculation.calculateTotalWidth400)(valueText, relativeValueText);
      }
      return nameWidth + valueWidth;
    })
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    SeriesWithVirtualizer,
    {
      data,
      activeItem,
      seriesItemTemplate,
      maxContentWidth: maxContentWidthWithinFirstTenValues
    }
  );
};
const SeriesWithVirtualizer = (props) => {
  const { data, seriesItemTemplate, activeItem, maxContentWidth } = props;
  const isItemSelected = (0, import_selectors.useSelector)(({ activeSeries }) => !!activeSeries);
  const parentRef = (0, import_react.useRef)(null);
  const config = (0, import_ExternalConfigContext.useExternalConfig)();
  const isSeriesHeaderDisabled = config.seriesHeader?.props?.disabled;
  const scrollbarRefCallback = (0, import_use_scrollbar_detection.useScrollbarDetection)();
  const combinedRefCallback = (0, import_react.useCallback)(
    (element) => {
      parentRef.current = element;
      scrollbarRefCallback(element);
    },
    [scrollbarRefCallback]
  );
  const virtualizer = (0, import_react_virtual.useVirtualizer)({
    count: data.length,
    getScrollElement: () => parentRef.current,
    estimateSize: (index) => (0, import_estimate_series_item_size.estimateSeriesItemSize)(data[index], maxContainerWidth),
    paddingStart: isSeriesHeaderDisabled ? 0 : 24,
    paddingEnd: 0,
    overscan: 3
  });
  const spacings = import_constants.TOOLTIP_WRAPPER_PADDING * 2 + import_constants.SYMBOL_AND_SPACING_WIDTH;
  const maxContainerWidth = maxContentWidth + spacings + import_constants.TOOLTIP_CONTENT_PADDING * 2;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { ref: combinedRefCallback, className: import_tooltip_css.seriesWrapper, children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    "div",
    {
      className: import_tooltip_css.seriesOverflowContainer,
      style: {
        height: `${virtualizer.getTotalSize()}px`,
        width: `${Math.max(maxContainerWidth, import_constants.MIN_TOOLTIP_WIDTH)}px`,
        maxWidth: "100%"
      },
      children: [
        config.seriesHeader,
        virtualizer.getVirtualItems().map((item) => {
          return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "div",
            {
              className: (0, import_clsx.default)(
                import_tooltip_css.seriesGridContainer,
                item.index === activeItem && !isItemSelected && import_tooltip_css.seriesItemHighlighted
              ),
              style: {
                position: "absolute",
                top: 0,
                left: 0,
                width: `calc(100% - var(--needs-scrollbar-fix, 0) * ${import_constants.SCROLLBAR_WIDTH}px)`,
                height: `${item.size}px`,
                transform: `translateY(${item.start}px)`
              },
              children: seriesItemTemplate(data[item.index])
            },
            item.key
          );
        })
      ]
    }
  ) });
};
