// @dynatrace/strato-components/esm/forms/number-input-v2/utils/calculations.js
import { parseNumber } from "./format.js";
import { sanitizePositiveNumber } from "./validation.js";
function resolveInputValue(value) {
  if (value === null) {
    return 0;
  }
  if (typeof value === "string") {
    if (value.trim() === "") {
      return 0;
    }
    const parsed = parseNumber(value);
    return parsed ?? 0;
  }
  return value;
}
function isValidStep(step) {
  return step !== void 0 && !Number.isNaN(step) && step > 0;
}
function computeStepAlignedValue(parsedValue, direction, cleanedStep, baseValue) {
  const remainder = Math.abs((parsedValue - baseValue) % cleanedStep);
  const softEpsilon = Number.EPSILON * 100;
  if (remainder <= softEpsilon || remainder >= cleanedStep - softEpsilon) {
    return void 0;
  }
  const nextValue = parsedValue + direction * cleanedStep;
  if (parsedValue < baseValue) {
    return nextValue + remainder - (direction > 0 ? cleanedStep : 0);
  }
  return nextValue - remainder + (direction < 0 ? cleanedStep : 0);
}
function applyMinConstraint(value, min) {
  if (min === void 0 || Number.isNaN(min)) {
    return value;
  }
  return Math.max(value, min);
}
function applyMaxConstraint(value, max, baseValue, cleanedStep, validStep) {
  if (max === void 0 || Number.isNaN(max)) {
    return value;
  }
  if (validStep && value >= max) {
    return baseValue + cleanedStep * Math.floor((max - baseValue) / cleanedStep);
  }
  return Math.min(value, max);
}
function roundToPrecision(value) {
  const precision = Math.pow(10, 10);
  return Math.round(value * precision) / precision;
}
function applyStep(value, step, stepMultiplier, min, max, direction) {
  const parsedValue = resolveInputValue(value);
  if (Number.isNaN(parsedValue)) {
    return value;
  }
  const validStep = isValidStep(step);
  const cleanedStep = sanitizePositiveNumber(step);
  const effectiveStepSize = Math.trunc(sanitizePositiveNumber(stepMultiplier)) * cleanedStep;
  const baseValue = min || 0;
  let nextValue = parsedValue + direction * effectiveStepSize;
  if (validStep) {
    const aligned = computeStepAlignedValue(
      parsedValue,
      direction,
      cleanedStep,
      baseValue
    );
    if (aligned !== void 0) {
      nextValue = aligned;
    }
  }
  nextValue = applyMinConstraint(nextValue, min);
  nextValue = applyMaxConstraint(
    nextValue,
    max,
    baseValue,
    cleanedStep,
    validStep
  );
  return roundToPrecision(nextValue);
}
export {
  applyStep
};
