// @dynatrace/strato-components/esm/forms/number-input-v2/utils/validation.js
import { parseNumber } from "./format.js";
import { getI18nErrorMessage } from "../../forms-core/hooks/useFormFieldValidation.js";
import {
  FormFieldMessagesAction
} from "../../forms-core/state/formfield-messages-state-reducer.js";
function sanitizePositiveNumber(value) {
  return value !== void 0 && !Number.isNaN(value) && value > 0 ? value : 1;
}
function isStepAligned(value, step, min = 0) {
  step = sanitizePositiveNumber(step);
  const roundedValue = Math.round((value - min) / step) * step + min;
  const tolerance = step * 1e-9;
  return Math.abs(value - roundedValue) < tolerance;
}
function addError(errorType, errorId, ref, dispatch) {
  dispatch?.({
    type: FormFieldMessagesAction.ADDMESSAGE,
    payload: {
      id: errorId,
      message: getI18nErrorMessage(errorType, ref),
      validityStateKey: errorType,
      messageType: "formcontrol",
      variant: "error"
    }
  });
}
function getErrorKey(value, min, max, step) {
  const parsed = parseNumber(value);
  if (min !== void 0 && parsed < min) {
    return "rangeUnderflow";
  }
  if (max !== void 0 && parsed > max) {
    return "rangeOverflow";
  }
  if (step && step > 0 && !isStepAligned(parsed, step, min ?? 0)) {
    return "stepMismatch";
  }
  return null;
}
export {
  addError,
  getErrorKey,
  isStepAligned,
  sanitizePositiveNumber
};
