// @dynatrace/strato-components/esm/tables/DataTable/features/LoadingState/LoadingState.js
import { jsx } from "react/jsx-runtime";
import { assignInlineVars } from "@vanilla-extract/dynamic";
import { useEffect, useRef, useState } from "react";
import { useIntl } from "react-intl";
import * as styles from "./LoadingState.sty.js";
import { ProgressCircle } from "../../../../content/progress/ProgressCircle.js";
import { Flex } from "../../../../layouts/flex/Flex.js";
import { dataTableMessages } from "../../messages.js";
const PROGRESS_CIRCLE_HEIGHT_BREAKPOINT = 120;
function useLoadingState(props, options) {
  options.state = { ...options.state, loading: props.loading };
}
function LoadingState(props) {
  const { headerHeight } = props;
  const loadingContainerRef = useRef(null);
  const [height, setHeight] = useState(void 0);
  const intl = useIntl();
  useEffect(() => {
    if (!loadingContainerRef.current) {
      return;
    }
    const resizeObserver = new ResizeObserver((entries) => {
      if (entries.length > 0) {
        setHeight(entries[0].contentRect.height);
      }
    });
    resizeObserver.observe(loadingContainerRef.current);
    return () => {
      resizeObserver.disconnect();
    };
  }, []);
  return /* @__PURE__ */ jsx(
    Flex,
    {
      ref: loadingContainerRef,
      justifyContent: "center",
      alignItems: "center",
      className: styles.loadingState,
      style: assignInlineVars({
        [styles.dataTableHeaderHeight]: `${headerHeight}px`
      }),
      children: /* @__PURE__ */ jsx(
        ProgressCircle,
        {
          size: height !== void 0 && height > PROGRESS_CIRCLE_HEIGHT_BREAKPOINT ? "large" : "small",
          "aria-label": intl.formatMessage(dataTableMessages.loadingData)
        }
      )
    }
  );
}
export {
  LoadingState,
  PROGRESS_CIRCLE_HEIGHT_BREAKPOINT,
  useLoadingState
};
