// @dynatrace/strato-components/charts/gantt-chart/context/GanttChartStateContext.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var GanttChartStateContext_exports = {};
__export(GanttChartStateContext_exports, {
  GanttChartStateContext: () => GanttChartStateContext,
  ganttChartStateInitialState: () => ganttChartStateInitialState
});
module.exports = __toCommonJS(GanttChartStateContext_exports);
var import_react = require("react");
var import_store = require("../../../core/store/store.js");
const ganttChartStateInitialState = {
  tooltip: {
    pinned: false,
    visible: false
  },
  overlay: {
    mouseInBounds: false
  }
};
const GanttChartStateContext = (0, import_react.createContext)(
  (0, import_store.createStore)(ganttChartStateInitialState)
);
