// @dynatrace/strato-components/esm/charts/annotations-chart/AnnotationsChart.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import { isEmpty } from "lodash-es";
import {
  forwardRef,
  useImperativeHandle,
  useRef
} from "react";
import { AnnotationsDisplay } from "./components/AnnotationsDisplay.js";
import { AnnotationsAxis } from "./components/axis/AnnotationsAxis.js";
import { ANNOTATIONS_CHART_ARIA_LABEL } from "./constants.js";
import { useAnnotationsChartConfig } from "./hooks/useAnnotationsChartConfig.js";
import { AnnotationsChartConfigProvider } from "./providers/annotations-chart-config.provider.js";
import { Marker } from "./slots/Marker.js";
import { Tooltip } from "./slots/Tooltip.js";
import { Track } from "./slots/Track.js";
import { XAxis } from "./slots/XAxis.js";
import { iterateSlots } from "./utils/iterate-slots.js";
import { buildScale } from "./utils/scales.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useMergeRefs } from "../../core/hooks/useMergeRefs.js";
import { defaultChartSize } from "../../core/styles/defaultChartValues.sty.js";
import { AnnotationsScrollWrapper } from "../core/components/annotations/components/overlays/AnnotationsScrollWrapper.js";
import {
  DEFAULT_LABELS_OFFSET,
  TRACK_HEIGHT,
  TRACK_SEPARATOR_STROKE_WIDTH
} from "../core/components/annotations/constants.js";
import { isNumericalData } from "../core/components/annotations/utils/is-numerical-marker.js";
import { ChartStateHandler } from "../core/components/chart-states/ChartStateHandler.js";
import { getChartStatesTemplates } from "../core/components/chart-states/utils/get-chart-states-template.js";
import { DataDtrumContext } from "../core/context/DataDtrumContext.js";
import { useObserveContainerSize } from "../core/hooks/useObserveContainerSize.js";
import { EmptyStateSlot } from "../core/slots/chart-states/EmptyStateSlot.js";
import { ErrorStateSlot } from "../core/slots/chart-states/ErrorStateSlot.js";
import { ChartType } from "../core/types/chart-types.js";
import { coerceSizeValue } from "../core/utils/coerceSizeValue.js";
const AXIS_HEIGHT = 25;
const X_AXIS_LABEL_HEIGHT = 10;
const COMPONENT_NAME = "AnnotationsChart";
const calculateXAxisHeight = (isHidden, label) => isHidden ? 0 : label ? AXIS_HEIGHT + X_AXIS_LABEL_HEIGHT : AXIS_HEIGHT;
const _AnnotationsChart = forwardRef((props, forwardedRef) => {
  const {
    annotationsActions,
    children,
    loading = false,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "data-testid": dataTestId,
    style: consumerStyle,
    className: consumerClassName,
    width: propsWidth
  } = props;
  const [behaviorTrackingProps] = useBehavioralTrackingProps(props);
  const [containerDimension, sizeObserverRef] = useObserveContainerSize();
  const { aggregatedMarkers, tracks, xAxisProps, tooltipProps } = iterateSlots(children);
  const { emptyState, errorState } = getChartStatesTemplates(children);
  const hasNumericalData = aggregatedMarkers.length > 0 && isNumericalData(aggregatedMarkers);
  const config = useAnnotationsChartConfig(
    props,
    xAxisProps,
    tooltipProps,
    hasNumericalData
  );
  const {
    showLabels,
    xAxisConfig,
    tooltipConfig,
    textOverflow,
    truncateMode,
    maxLines
  } = config;
  const shouldShowLabels = tracks.some(({ label }) => label) && showLabels;
  const labelsOffset = shouldShowLabels ? DEFAULT_LABELS_OFFSET + config.labelWidth : 0;
  const scale = buildScale(
    aggregatedMarkers,
    containerDimension.width - labelsOffset,
    xAxisConfig
  );
  const shownTracks = tracks.filter(({ hidden }) => !hidden);
  const tracksHeight = shownTracks.length * TRACK_HEIGHT + TRACK_SEPARATOR_STROKE_WIDTH;
  const axisHeight = calculateXAxisHeight(
    xAxisConfig.hidden,
    config.xAxisConfig.label
  );
  const coercedHeight = coerceSizeValue(config.height);
  const coercedWidth = coerceSizeValue(propsWidth);
  const height = coercedHeight ?? tracksHeight + axisHeight;
  const width = coercedWidth ?? "100%";
  const scrollWrapperHeight = Math.min(
    tracksHeight,
    Math.max(0, containerDimension.height - axisHeight)
  );
  const containerRef = useRef(null);
  const rootRef = useMergeRefs([
    forwardedRef,
    containerRef,
    sizeObserverRef
  ]);
  useImperativeHandle(forwardedRef, () => ({
    element: containerRef.current,
    getConfig: () => {
      return JSON.stringify({
        type: ChartType.AnnotationsChart,
        ...config
      });
    }
  }));
  return /* @__PURE__ */ jsx(
    ChartStateHandler,
    {
      empty: isEmpty(tracks),
      height: "100%",
      emptyState,
      errorState,
      isLoading: loading,
      className: clsx(defaultChartSize, consumerClassName),
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "data-testid": dataTestId,
      "data-dt-component": COMPONENT_NAME,
      style: consumerStyle,
      children: /* @__PURE__ */ jsx(
        "div",
        {
          style: { height, width },
          ref: rootRef,
          "aria-label": ANNOTATIONS_CHART_ARIA_LABEL,
          role: "img",
          children: /* @__PURE__ */ jsxs(AnnotationsChartConfigProvider, { config, children: [
            /* @__PURE__ */ jsx(DataDtrumContext.Provider, { value: { dataDtrumAllow, dataDtrumMask }, children: /* @__PURE__ */ jsx(
              AnnotationsScrollWrapper,
              {
                height: scrollWrapperHeight,
                width: containerDimension.width,
                children: /* @__PURE__ */ jsx(
                  AnnotationsDisplay,
                  {
                    tracks,
                    scale,
                    annotationsActions,
                    showLabels: shouldShowLabels,
                    labelsOffset,
                    width: containerDimension.width,
                    tooltipConfig,
                    textOverflow,
                    truncateMode,
                    maxLines,
                    ...behaviorTrackingProps
                  }
                )
              }
            ) }),
            /* @__PURE__ */ jsx(
              AnnotationsAxis,
              {
                height: axisHeight,
                scale,
                width: containerDimension.width,
                labelsOffset
              }
            )
          ] })
        }
      )
    }
  );
});
const AnnotationsChart = Object.assign(_AnnotationsChart, {
  Track,
  Marker,
  XAxis,
  Tooltip,
  EmptyState: EmptyStateSlot,
  ErrorState: ErrorStateSlot
});
export {
  AnnotationsChart,
  _AnnotationsChart
};
