// @dynatrace/strato-components/esm/tables/DataTable/components/Toolbar/ColumnSettings/ColumnOrderSettings/column-order-settings-dnd-utils.js
import Animations from "@dynatrace/strato-design-tokens/animations";
import * as columnSettingsDndStyles from "./ColumnOrderSettingsDragAndDrop.sty.js";
import { getIsColumnPinned } from "../../../../features/ColumnPinning/column-pinning-utils.js";
import * as columnListItemStyles from "../DataTableColumnsListItem.sty.js";
const DATA_TABLE_COLUMNS_LIST_ITEM_BOTTOM_BORDER = 1;
const DATA_TABLE_COLUMNS_LIST_ITEM_MIN_HEIGHT = 32 + DATA_TABLE_COLUMNS_LIST_ITEM_BOTTOM_BORDER;
let lastAffectedItems = /* @__PURE__ */ new Set();
let hiddenChildColumns = /* @__PURE__ */ new Set();
function transformColumnsOnDragOver(columnsList, event, localColumnOrder, transformValue, activeColumn) {
  if (!columnsList || !activeColumn) {
    return;
  }
  const { over, collisions } = event;
  if (!over || !collisions?.length) {
    return;
  }
  const newAffectedColumns = /* @__PURE__ */ new Set();
  const { fromIndex, toIndex } = getDragIndexes(
    localColumnOrder,
    activeColumn.id,
    over.data.current?.columnId
  );
  const adjustedTransformValue = transformValue + DATA_TABLE_COLUMNS_LIST_ITEM_BOTTOM_BORDER;
  if (fromIndex < toIndex) {
    const adaptedFromIndex = fromIndex + activeColumn.columns.length;
    transformColumns(
      columnsList,
      localColumnOrder.slice(adaptedFromIndex + 1, toIndex + 1),
      -adjustedTransformValue,
      newAffectedColumns
    );
  } else if (fromIndex > toIndex) {
    transformColumns(
      columnsList,
      localColumnOrder.slice(toIndex, fromIndex),
      adjustedTransformValue,
      newAffectedColumns
    );
  }
  transformColumns(columnsList, lastAffectedItems, 0);
  lastAffectedItems = newAffectedColumns;
}
function getColumnById(columnsList, columnId) {
  return columnsList.querySelector(
    `[data-columnid="${CSS.escape(columnId)}"]:not([data-pinned-position])`
  );
}
function getAllColumns(columnsList) {
  const allLocalColumns = columnsList.querySelectorAll("[data-columnid]");
  return Array.from(allLocalColumns);
}
function getAllNonTargetColumns(columnsList, activeParentId) {
  const selector = activeParentId ? `[data-pinned-position], [data-columnid]:not([data-parentcolumnid="${CSS.escape(
    String(activeParentId)
  )}"])` : "[data-pinned-position]";
  const nonTargetColumns = columnsList.querySelectorAll(selector);
  return Array.from(nonTargetColumns);
}
function translateColumnY(localColumn, value) {
  const transform = `translateY(${value}px)`;
  const transition = `transform ${Animations.SlideIn.Duration.Base} ${Animations.Move.Easing}`;
  localColumn.style.transform = transform;
  localColumn.style.transition = transition;
}
function resetAllColumnsStyles(localColumns) {
  for (const column of localColumns) {
    column.style.removeProperty("transform");
    column.style.removeProperty("transition");
    column.classList.remove(
      columnSettingsDndStyles.disabledNonTargetLocalColumn,
      columnListItemStyles.dragging
    );
  }
  lastAffectedItems.clear();
  hiddenChildColumns.clear();
}
function styleNonTargetColumns(columnsList, activeParentId) {
  const nonTargetLocalColumns = getAllNonTargetColumns(
    columnsList,
    activeParentId
  );
  for (const column of nonTargetLocalColumns) {
    column.classList.add(columnSettingsDndStyles.disabledNonTargetLocalColumn);
  }
}
function styleDraggedChildColumns(columnsList, childColumnIds) {
  for (const columnId of childColumnIds) {
    const childColumn = getColumnById(columnsList, columnId);
    if (childColumn) {
      childColumn.classList.add(columnListItemStyles.dragging);
    }
  }
}
function hideChildColumns(columnsList, childColumnIds) {
  hiddenChildColumns = new Set(childColumnIds);
  styleDraggedChildColumns(columnsList, childColumnIds);
}
function hideChildColumnsOnScroll(columnsList) {
  if (hiddenChildColumns.size === 0) {
    return;
  }
  styleDraggedChildColumns(columnsList, hiddenChildColumns);
}
function getDragIndexes(localColumnOrder, activeId, overId) {
  const fromIndex = localColumnOrder.indexOf(activeId);
  const toIndex = localColumnOrder.indexOf(overId);
  return { fromIndex, toIndex };
}
function transformColumns(container, columnIds, amountToTransform, newAffectedItems) {
  for (const columnId of columnIds) {
    const column = getColumnById(container, columnId);
    if (!column) {
      continue;
    }
    translateColumnY(column, amountToTransform);
    if (newAffectedItems) {
      newAffectedItems.add(columnId);
      lastAffectedItems.delete(columnId);
    }
  }
}
const DATA_TABLE_COLUMNS_LIST_DND_COLLISION_OFFSET = DATA_TABLE_COLUMNS_LIST_ITEM_MIN_HEIGHT / 2;
function getLeafColumnsForColumn(column, localColumnOrder) {
  if (column.columns.length === 0) {
    return [];
  }
  return localColumnOrder.filter(
    (columnId) => column.columns.some((col) => col.id === columnId)
  );
}
function closestTargetWithinY({
  droppableContainers,
  droppableRects,
  collisionRect,
  active
}) {
  const initialRect = active.rect.current.initial;
  if (initialRect === null) {
    return [];
  }
  const collisions = [];
  const isDraggingUp = collisionRect.top <= initialRect.top;
  const collisionRectY = isDraggingUp ? collisionRect.top + DATA_TABLE_COLUMNS_LIST_DND_COLLISION_OFFSET : collisionRect.bottom - DATA_TABLE_COLUMNS_LIST_DND_COLLISION_OFFSET;
  for (const droppableContainer of droppableContainers) {
    const { id } = droppableContainer;
    const rect = droppableRects.get(id);
    if (rect === void 0) {
      continue;
    }
    if (rect.top <= collisionRectY && collisionRectY <= rect.bottom) {
      return [
        {
          id,
          data: { droppableContainer }
        }
      ];
    }
    const droppableContainerY = isDraggingUp ? rect.top : rect.bottom;
    const distance = Math.abs(collisionRectY - droppableContainerY);
    collisions.push({ id, data: { droppableContainer, value: distance } });
  }
  return collisions.toSorted((a, b) => a.data.value - b.data.value);
}
function createColumnOrderSettingsCollisionDetection(localColumnOrder, localColumnPinning, draggingColumnInfo) {
  return ({ active, droppableContainers, ...rest }) => {
    if (!draggingColumnInfo) {
      return [];
    }
    const { column: activeColumn, index: activeIndex } = draggingColumnInfo;
    if (activeColumn === void 0) {
      return [];
    }
    const activeParentId = activeColumn.parent?.id;
    const validTargets = droppableContainers.filter((target) => {
      const currentTarget = target.data.current;
      const targetColumn = currentTarget?.column;
      const targetParentColumn = targetColumn.parent;
      const targetIndex = currentTarget?.columnIndex;
      if (activeIndex === targetIndex) {
        return true;
      }
      const hasSameParent = targetParentColumn?.id === activeParentId;
      if (activeParentId !== void 0) {
        return hasSameParent;
      }
      if (targetColumn.columns.length > 0) {
        return activeIndex > targetIndex;
      }
      if (targetParentColumn?.id !== void 0) {
        const orderedChildIds = getLeafColumnsForColumn(
          targetParentColumn,
          localColumnOrder
        );
        const lastUnpinnedChildId = orderedChildIds.findLast(
          (childId) => !getIsColumnPinned(childId, localColumnPinning)
        );
        if (targetColumn.id === lastUnpinnedChildId) {
          return activeIndex < targetIndex;
        }
      }
      return hasSameParent;
    });
    return closestTargetWithinY({
      active,
      droppableContainers: validTargets,
      ...rest
    });
  };
}
function restrictToVerticalBounds(transform, rect, verticalBounds) {
  const value = {
    ...transform
  };
  if (rect.top + transform.y <= verticalBounds.top) {
    value.y = verticalBounds.top - rect.top;
  } else if (rect.bottom + transform.y >= verticalBounds.bottom) {
    value.y = verticalBounds.bottom - rect.bottom;
  }
  return value;
}
function createRestrictToColumnsListAreaModifier(modalContent) {
  return ({ transform, draggingNodeRect }) => {
    if (!draggingNodeRect || modalContent === null) {
      return transform;
    }
    transform.x = 0;
    const columnsListRect = modalContent.getBoundingClientRect();
    const verticalBounds = {
      top: columnsListRect.top,
      bottom: columnsListRect.bottom
    };
    const lastLeftPinned = Array.from(
      modalContent.querySelectorAll('[data-pinned-position="left"]')
    ).at(-1) ?? null;
    const firstRightPinned = modalContent.querySelector(
      '[data-pinned-position="right"]'
    );
    let adjustedTopBound = verticalBounds.top;
    if (lastLeftPinned) {
      const pinnedBottom = lastLeftPinned.getBoundingClientRect().bottom;
      adjustedTopBound = Math.max(adjustedTopBound, pinnedBottom);
    }
    verticalBounds.top = adjustedTopBound;
    if (firstRightPinned) {
      verticalBounds.bottom = Math.min(
        verticalBounds.bottom,
        firstRightPinned.getBoundingClientRect().top - // account for the bottom border of the last item
        DATA_TABLE_COLUMNS_LIST_ITEM_BOTTOM_BORDER
      );
    }
    return restrictToVerticalBounds(
      transform,
      draggingNodeRect,
      verticalBounds
    );
  };
}
function getDragHandleForColumn(columnId, columnsList) {
  if (columnsList === null) {
    return null;
  }
  return columnsList?.querySelector(
    `[data-columnid="${CSS.escape(
      columnId
    )}"]:not([data-pinned-position]) button`
  ) ?? null;
}
export {
  DATA_TABLE_COLUMNS_LIST_ITEM_BOTTOM_BORDER,
  DATA_TABLE_COLUMNS_LIST_ITEM_MIN_HEIGHT,
  createColumnOrderSettingsCollisionDetection,
  createRestrictToColumnsListAreaModifier,
  getAllColumns,
  getColumnById,
  getDragHandleForColumn,
  getDragIndexes,
  getLeafColumnsForColumn,
  hideChildColumns,
  hideChildColumnsOnScroll,
  resetAllColumnsStyles,
  styleNonTargetColumns,
  transformColumnsOnDragOver
};
