// @dynatrace/strato-components/esm/tables/DataTable/components/Toolbar/ColumnSettings/ColumnOrderSettings/ColumnOrderSettingsDragAndDrop.js
import { jsx, jsxs } from "react/jsx-runtime";
import {
  DndContext as DragAndDropContext,
  DragOverlay,
  KeyboardSensor,
  MouseSensor,
  TouchSensor,
  useSensor,
  useSensors
} from "@dnd-kit/core";
import clsx from "clsx";
import { useCallback, useMemo, useRef, useState } from "react";
import {
  createColumnOrderSettingsCollisionDetection,
  createRestrictToColumnsListAreaModifier,
  DATA_TABLE_COLUMNS_LIST_ITEM_BOTTOM_BORDER,
  DATA_TABLE_COLUMNS_LIST_ITEM_MIN_HEIGHT,
  getAllColumns,
  getColumnById,
  getDragHandleForColumn,
  getDragIndexes,
  getLeafColumnsForColumn,
  hideChildColumns,
  hideChildColumnsOnScroll,
  resetAllColumnsStyles,
  styleNonTargetColumns,
  transformColumnsOnDragOver
} from "./column-order-settings-dnd-utils.js";
import * as styles from "./ColumnOrderSettingsDragAndDrop.sty.js";
import { devLog } from "../../../../../../core/utils/logging.js";
const SCROLL_JUMP_THRESHOLD = DATA_TABLE_COLUMNS_LIST_ITEM_MIN_HEIGHT * 10;
function ColumnOrderSettingsDragAndDropContext({
  children,
  localColumnOrder,
  updateLocalColumnOrder,
  localColumnPinning,
  itemVirtualizer
}) {
  const [dragOverlayContent, setDragOverlayContent] = useState(null);
  const [dragOverlaySize, setDragOverlaySize] = useState({ width: 0, height: 0 });
  const [draggingColumnInfo, setDraggingColumnInfo] = useState(null);
  const modalContentRef = useRef(null);
  const scrollHandlerRef = useRef(null);
  const sensors = useSensors(
    useSensor(MouseSensor, {}),
    useSensor(TouchSensor, {}),
    useSensor(KeyboardSensor, {})
  );
  const columnsListCollisionDetection = useMemo(
    () => createColumnOrderSettingsCollisionDetection(
      localColumnOrder,
      localColumnPinning,
      draggingColumnInfo
    ),
    [draggingColumnInfo, localColumnOrder, localColumnPinning]
  );
  const restrictToColumnsListAreaModifier = useCallback(
    (args) => {
      return createRestrictToColumnsListAreaModifier(
        itemVirtualizer.scrollElement
      )(args);
    },
    [itemVirtualizer]
  );
  const resetDragging = () => {
    const scrollElement = itemVirtualizer.scrollElement;
    if (scrollElement && scrollHandlerRef.current) {
      scrollElement.removeEventListener("scroll", scrollHandlerRef.current);
      scrollHandlerRef.current = null;
    }
    if (modalContentRef.current) {
      const allColumns = getAllColumns(modalContentRef.current);
      resetAllColumnsStyles(allColumns);
    }
    modalContentRef.current = null;
    setDraggingColumnInfo(null);
    setDragOverlayContent(null);
    setDragOverlaySize({ width: 0, height: 0 });
  };
  const resetDraggingAndFocusLastActive = (activeId) => {
    requestAnimationFrame(() => {
      getDragHandleForColumn(activeId, modalContentRef.current)?.focus();
      resetDragging();
    });
  };
  const handleOnDragStart = ({ active }) => {
    const activeLocalColumn = active.data.current?.column;
    if (!activeLocalColumn) {
      return;
    }
    const scrollElement = itemVirtualizer.scrollElement;
    modalContentRef.current = scrollElement?.firstElementChild || scrollElement;
    const localColumn = getColumnById(
      modalContentRef.current,
      activeLocalColumn.id
    );
    if (!localColumn) {
      devLog(
        "[DataTable] Could not find column element in DOM",
        activeLocalColumn.id
      );
      return;
    }
    const columnInfo = {
      id: active.id,
      index: active.data.current?.columnIndex,
      column: activeLocalColumn
    };
    setDraggingColumnInfo(columnInfo);
    let lastScrollOffset = itemVirtualizer.scrollOffset ?? scrollElement?.scrollTop ?? 0;
    const handleScroll = () => {
      const innerContainer = modalContentRef.current;
      if (!innerContainer || !columnInfo.column) {
        return;
      }
      styleNonTargetColumns(innerContainer, columnInfo.column.parent?.id);
      hideChildColumnsOnScroll(innerContainer);
      const currentOffset = scrollElement.scrollTop;
      const scrollDelta = Math.abs(currentOffset - lastScrollOffset);
      if (scrollDelta >= SCROLL_JUMP_THRESHOLD) {
        scrollElement.scrollTop = lastScrollOffset;
        return;
      }
      lastScrollOffset = currentOffset;
    };
    scrollHandlerRef.current = handleScroll;
    scrollElement?.addEventListener("scroll", handleScroll);
    styleNonTargetColumns(
      modalContentRef.current,
      activeLocalColumn.parent?.id
    );
    const dragOverlaySize2 = {
      width: localColumn.clientWidth,
      height: localColumn.clientHeight
    };
    const clonedElement = localColumn.cloneNode(true);
    if (activeLocalColumn.columns.length === 0) {
      setDragOverlayContent(clonedElement);
    } else {
      const wrapper = document.createElement("div");
      wrapper.appendChild(clonedElement);
      const childColumnIds = getLeafColumnsForColumn(
        activeLocalColumn,
        localColumnOrder
      );
      for (const childColumnId of childColumnIds) {
        const childColumn = getColumnById(
          modalContentRef.current,
          childColumnId
        );
        if (childColumn) {
          const clonedChild = childColumn.cloneNode(true);
          wrapper.appendChild(clonedChild);
          dragOverlaySize2.height += childColumn.clientHeight + DATA_TABLE_COLUMNS_LIST_ITEM_BOTTOM_BORDER;
        }
      }
      hideChildColumns(modalContentRef.current, childColumnIds);
      setDragOverlayContent(wrapper);
    }
    setDragOverlaySize(dragOverlaySize2);
  };
  const handleOnDragOver = (event) => {
    transformColumnsOnDragOver(
      modalContentRef.current,
      event,
      localColumnOrder,
      dragOverlaySize.height,
      draggingColumnInfo?.column
    );
  };
  const handleOnDragEnd = ({ over }) => {
    if (!draggingColumnInfo) {
      devLog("[Table] On drag end: dragging column info is undefined");
      resetDragging();
      return;
    }
    const activeColumn = draggingColumnInfo.column;
    const activeId = activeColumn.id;
    if (over === void 0 || over === null) {
      devLog("[Table] On drag end: over info is undefined or null", over);
      resetDraggingAndFocusLastActive(activeId);
      return;
    }
    const overId = over.data.current?.columnId;
    if (overId === activeId) {
      resetDraggingAndFocusLastActive(activeId);
      return;
    }
    const { fromIndex, toIndex } = getDragIndexes(
      localColumnOrder,
      activeId,
      overId
    );
    const columnsToMove = [activeId];
    if (activeColumn.columns.length > 0) {
      const childColumnIds = getLeafColumnsForColumn(
        activeColumn,
        localColumnOrder
      );
      columnsToMove.push(...childColumnIds);
    }
    const directionOffset = fromIndex < toIndex ? 1 : -1;
    const adjustedTargetIndex = fromIndex < toIndex ? Math.max(0, toIndex - columnsToMove.length + directionOffset) : toIndex;
    const newColumnOrder = [...localColumnOrder];
    newColumnOrder.splice(fromIndex, columnsToMove.length);
    newColumnOrder.splice(adjustedTargetIndex, 0, ...columnsToMove);
    updateLocalColumnOrder(newColumnOrder);
    resetDraggingAndFocusLastActive(activeId);
  };
  const onDragCancel = () => {
    const activeId = draggingColumnInfo?.id;
    if (activeId) {
      resetDraggingAndFocusLastActive(activeId);
    }
  };
  return /* @__PURE__ */ jsxs(
    DragAndDropContext,
    {
      accessibility: { restoreFocus: false },
      sensors,
      collisionDetection: columnsListCollisionDetection,
      modifiers: [restrictToColumnsListAreaModifier],
      onDragStart: handleOnDragStart,
      onDragOver: handleOnDragOver,
      onDragEnd: handleOnDragEnd,
      onDragCancel,
      children: [
        children,
        /* @__PURE__ */ jsx(
          DragOverlay,
          {
            dropAnimation: null,
            className: clsx(styles.columnDndOverlay),
            style: {
              height: dragOverlaySize.height,
              width: dragOverlaySize.width
            },
            children: /* @__PURE__ */ jsx(
              "div",
              {
                ref: (node) => dragOverlayContent && node?.appendChild(dragOverlayContent)
              }
            )
          }
        )
      ]
    }
  );
}
export {
  ColumnOrderSettingsDragAndDropContext
};
