// @dynatrace/strato-components/esm/content/information-overlay/InformationOverlay.js
import { jsx, jsxs } from "react/jsx-runtime";
import { forwardRef, useMemo } from "react";
import { Content } from "./Content.js";
import { InformationOverlayContext } from "./contexts/InformationOverlayContext.js";
import { COMPONENT_NAME } from "./information-overlay-constants.js";
import { InformationOverlayIcon } from "./InformationOverlayIcon.js";
import { Trigger } from "./Trigger.js";
import { useAriaLabelingProps } from "../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useControllableState } from "../../core/hooks/useControllableState.js";
import { useId } from "../../core/hooks/useId.js";
import { useMergeRefs } from "../../core/hooks/useMergeRefs.js";
import { useSafeForwardProps } from "../../core/hooks/useSafeForwardProps.js";
import { useSlots } from "../../core/hooks/useSlots.js";
import { Flex } from "../../layouts/flex/Flex.js";
import { useBreakpoint } from "../../layouts/hooks/useBreakpoint.js";
import { useOverlayWithTrigger } from "../../overlays/hooks/useOverlayWithTrigger.js";
import { Overlay } from "../../overlays/overlay/Overlay.js";
const InformationOverlayRoot = forwardRef((props, forwardedRef) => {
  const {
    children,
    defaultOpen,
    open,
    onOpenChange,
    color = "neutral",
    placement = "bottom",
    style: consumerStyle,
    className: consumerClassName,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
  const ariaLabelingProps = useAriaLabelingProps(restProps);
  const forwardProps = useSafeForwardProps(restProps, ariaLabelingProps);
  const overlayContentId = useId();
  const [isOpen, setIsOpen] = useControllableState({
    defaultValue: defaultOpen,
    value: open,
    onChange: onOpenChange
  });
  const matchesMaxWidthWindow = useBreakpoint("(max-width: 480px)");
  const {
    targetRef,
    triggerRef,
    overlayRef,
    overlayState,
    overlayProps,
    overlayTriggerProps
  } = useOverlayWithTrigger({
    placement: `${placement}-start`,
    placementFallbacks: [
      "top-start",
      "bottom-start",
      "right-start",
      "left-start"
    ],
    offset: 4,
    crossOffset: -16,
    widthStrategy: matchesMaxWidthWindow ? { max: 300 } : { max: 400 },
    open: isOpen,
    onOpenChange: setIsOpen,
    shouldCloseOnBlur: true
  });
  const overlayStateIsOpen = overlayState.isOpen;
  const rootRef = useMergeRefs([forwardedRef, triggerRef, targetRef]);
  const [slots] = useSlots(children, {
    trigger: Trigger,
    content: Content
  });
  const { children: triggerChildren, ...triggerProps } = slots.trigger?.props ?? {};
  const contextProviderValue = useMemo(() => {
    return { color, overlayTriggerProps, id: overlayContentId, isOpen };
  }, [color, isOpen, overlayContentId, overlayTriggerProps]);
  return /* @__PURE__ */ jsxs(InformationOverlayContext.Provider, { value: contextProviderValue, children: [
    /* @__PURE__ */ jsx(
      Trigger,
      {
        ref: rootRef,
        id: triggerProps.id,
        className: triggerProps.className,
        style: triggerProps.style,
        "aria-label": triggerProps["aria-label"],
        "aria-describedby": overlayStateIsOpen ? overlayContentId : void 0,
        "data-testid": triggerProps["data-testid"],
        "data-dtrum-mask": triggerProps["data-dtrum-mask"] ?? dataDtrumMask,
        "data-dtrum-allow": triggerProps["data-dtrum-allow"] ?? dataDtrumAllow,
        ...triggerProps,
        children: triggerChildren
      }
    ),
    /* @__PURE__ */ jsx(
      Overlay,
      {
        ref: overlayRef,
        isOpen: overlayStateIsOpen,
        overlayProps,
        className: consumerClassName,
        style: consumerStyle,
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        ...ariaLabelingProps,
        ...behavioralTrackingProps,
        ...forwardProps,
        "data-dt-component": COMPONENT_NAME,
        children: /* @__PURE__ */ jsx(
          Flex,
          {
            id: overlayContentId,
            paddingY: 12,
            paddingX: 16,
            flexDirection: "column",
            children: slots.content
          }
        )
      }
    )
  ] });
});
InformationOverlayRoot.displayName = COMPONENT_NAME;
const InformationOverlay = Object.assign(InformationOverlayRoot, {
  Content,
  Trigger,
  Icon: InformationOverlayIcon
});
export {
  InformationOverlay
};
