// @dynatrace/strato-components/esm/content/information-overlay/Trigger.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import { forwardRef } from "react";
import { useIntl } from "react-intl";
import {
  CriticalIcon,
  InformationIcon,
  SuccessIcon,
  WarningIcon
} from "@dynatrace/strato-icons";
import { useInformationOverlayContext } from "./hooks/useInformationOverlayContext.js";
import { COMPONENT_NAME } from "./information-overlay-constants.js";
import {
  InformationOverlayTriggerCSS,
  InformationOverlayTriggerIconCSS,
  InformationOverlayTriggerIconHoverCSS,
  InformationOverlayTriggerIconVariantsCSS,
  InformationOverlayTriggerTextCSS
} from "./InformationOverlay.sty.js";
import { InformationOverlayIcon } from "./InformationOverlayIcon.js";
import { useAriaLabelingProps } from "../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useId } from "../../core/hooks/useId.js";
import { useSlots } from "../../core/hooks/useSlots.js";
import { useFocusRing } from "../../core/styles/useFocusRing.js";
import { canReceiveDefaultTextFormat as _canReceiveDefaultTextFormat } from "../../core/utils/can-receive-default-text-format.js";
import { mergeProps as _mergeProps } from "../../core/utils/merge-props.js";
const Trigger = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const {
    children,
    id: propId,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    ...remainingProps
  } = props;
  const { color, overlayTriggerProps } = useInformationOverlayContext();
  const id = useId();
  const { focusProps, focusClassName, isFocusVisible } = useFocusRing({
    color: children ? "neutral" : color,
    isMinimal: true
  });
  const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
  const {
    "data-dt-name": dataDtName = "Trigger",
    ...remainingBehavioralTrackingProps
  } = behavioralTrackingProps;
  const { "aria-label": ariaLabelProp, ...remainingAriaLabelingProps } = useAriaLabelingProps(restProps);
  const [slots, freeform] = useSlots(children, {
    icon: InformationOverlayIcon
  });
  const hasFreeform = freeform.filter(
    (child) => child !== "" && child !== void 0 && child !== null
  ).length > 0;
  if (slots.icon === void 0) {
    switch (color) {
      case "success":
        slots.icon = /* @__PURE__ */ jsx(SuccessIcon, {});
        break;
      case "warning":
        slots.icon = /* @__PURE__ */ jsx(WarningIcon, {});
        break;
      case "critical":
        slots.icon = /* @__PURE__ */ jsx(CriticalIcon, {});
        break;
      default:
        slots.icon = /* @__PURE__ */ jsx(InformationIcon, {});
        break;
    }
  }
  const intl = useIntl();
  let ariaLabel = ariaLabelProp;
  if (ariaLabel === void 0 && (children === void 0 || !_canReceiveDefaultTextFormat(children))) {
    ariaLabel = intl.formatMessage({
      defaultMessage: "Open additional information.",
      id: "NXdoElYQZNKea1Sc",
      description: "Aria label fallback for Information overlay trigger"
    });
  }
  return /* @__PURE__ */ jsxs(
    "button",
    {
      ref: forwardedRef,
      "aria-haspopup": "dialog",
      "aria-label": ariaLabel,
      id: propId ?? id,
      ..._mergeProps(
        overlayTriggerProps,
        focusProps,
        remainingAriaLabelingProps,
        remainingBehavioralTrackingProps
      ),
      className: clsx(InformationOverlayTriggerCSS, consumerClassName),
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "data-dt-name": dataDtName,
      "data-dt-component": COMPONENT_NAME,
      children: [
        /* @__PURE__ */ jsx(
          InformationOverlayIcon,
          {
            className: clsx(
              InformationOverlayTriggerIconCSS,
              InformationOverlayTriggerIconVariantsCSS[color],
              {
                [focusClassName]: !hasFreeform,
                [InformationOverlayTriggerIconHoverCSS]: !hasFreeform
              },
              { "strato-focus-visible": isFocusVisible }
            ),
            children: slots.icon
          }
        ),
        hasFreeform && /* @__PURE__ */ jsx(
          "span",
          {
            className: clsx(focusClassName, InformationOverlayTriggerTextCSS, {
              "strato-focus-visible": isFocusVisible
            }),
            style: {
              flex: 1,
              overflow: "hidden",
              textDecoration: isFocusVisible ? "none" : void 0
            },
            children: freeform
          }
        )
      ]
    }
  );
});
Trigger.displayName = "InformationOverlay.Trigger";
export {
  Trigger
};
