// @dynatrace/strato-components/filters/filter/contexts/FilterBarContext.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var FilterBarContext_exports = {};
__export(FilterBarContext_exports, {
  FilterBarContext: () => FilterBarContext
});
module.exports = __toCommonJS(FilterBarContext_exports);
var import_react = require("react");
const initialState = { filters: {}, itemsPinnedState: {} };
const FilterBarContext = (0, import_react.createContext)({
  state: initialState,
  showLabels: true,
  lastAddedItem: null,
  setLastAddedItem: (value) => null,
  dispatch: () => null,
  hasControlledPinnedState: false
});
