// @dynatrace/strato-components/esm/charts/timeseries/components/displays/bar/accessors.js
import { isSymlogScale } from "../../../../core/utils/detect-scale.js";
import { safeGet } from "../../../../core/utils/safe-get.js";
import { getYScaleByUnit } from "../../../utils/yAxis.js";
const TIMESTAMP_BAR_WIDTH = 5;
const TIMESTAMP_BAR_X_DEVIATION = 2;
const getYCoordinates = ({ value = 0, cumulativeValue = 0, unit }, scale) => {
  const yScale = getYScaleByUnit(unit, scale);
  const valueCoordinate = yScale(value);
  const zeroCoordinate = yScale(0);
  const cumulativeCoordinate = yScale(cumulativeValue);
  const baseCumulative = yScale(cumulativeValue - value);
  const maxValue = yScale.invert(0);
  const maxValueCoordinate = yScale(maxValue);
  return {
    yScale,
    valueCoordinate,
    zeroCoordinate,
    cumulativeCoordinate,
    baseCumulative,
    maxValueCoordinate
  };
};
function getBarXPosition({ end, start }, xScale) {
  const isTimeframe = end && end !== start;
  return xScale(start) - (isTimeframe ? 0 : TIMESTAMP_BAR_X_DEVIATION);
}
function getBarYPosition(dataPointMetadata, yScale) {
  if (!yScale) {
    return 0;
  }
  const { valueCoordinate, zeroCoordinate, cumulativeCoordinate } = getYCoordinates(dataPointMetadata, yScale);
  let y2 = cumulativeCoordinate;
  if ((dataPointMetadata.value ?? 0) < 0) {
    const height2 = Math.abs(valueCoordinate - zeroCoordinate);
    y2 -= height2;
  }
  return y2;
}
function getBarHeight(dataPointMetadata, yScale) {
  const {
    valueCoordinate,
    zeroCoordinate,
    cumulativeCoordinate,
    baseCumulative,
    yScale: scale
  } = getYCoordinates(dataPointMetadata, yScale);
  let height2 = zeroCoordinate - valueCoordinate;
  const isLogValue = isSymlogScale(scale) && dataPointMetadata.cumulativeValue !== dataPointMetadata.value;
  if (isLogValue) {
    height2 = baseCumulative - cumulativeCoordinate;
  }
  return height2;
}
function getBarWidth({ end, start }, xScale) {
  const isTimeframe = end && end !== start;
  const width2 = isTimeframe ? xScale(end) - xScale(start) : TIMESTAMP_BAR_WIDTH;
  return width2;
}
function height(datapointsMap, yScale) {
  return (dataPoint) => {
    const dataPointMetadata = safeGet(datapointsMap, dataPoint);
    return getBarHeight(dataPointMetadata, yScale);
  };
}
function width(datapointsMap, xScale) {
  return (dataPoint) => {
    const dataPointMetadata = safeGet(datapointsMap, dataPoint);
    return getBarWidth(dataPointMetadata, xScale);
  };
}
function y(datapointsMap, yScale) {
  return (dataPoint) => {
    const dataPointMetadata = safeGet(datapointsMap, dataPoint);
    const { cumulativeCoordinate, maxValueCoordinate } = getYCoordinates(
      dataPointMetadata,
      yScale
    );
    const yCoordinate = maxValueCoordinate + cumulativeCoordinate;
    return yCoordinate;
  };
}
function x(datapointsMap, xScale) {
  return (dataPoint) => {
    const dataPointMetadata = safeGet(datapointsMap, dataPoint);
    return getBarXPosition(dataPointMetadata, xScale);
  };
}
export {
  getBarHeight,
  getBarWidth,
  getBarXPosition,
  getBarYPosition,
  height,
  width,
  x,
  y
};
