// @dynatrace/strato-components/esm/charts/timeseries/components/displays/bar/BarChartDisplay.js
import { jsx } from "react/jsx-runtime";
import { height, width, x, y } from "./accessors.js";
import { BarCanvas } from "../../../../core/components/canvas/components/BarCanvas.js";
import { useGridPlotSize } from "../../../../core/components/chart-grid/hooks/useGridPlotSize.js";
import { DEFAULT_FILL_OPACITY } from "../../../../core/constants.js";
import { safeGet } from "../../../../core/utils/safe-get.js";
import { BAR_CHART_DISPLAY_ARIA_LABEL } from "../../../constants.js";
import { useTimeseriesFilteredData } from "../../../hooks/useTimeseriesFilteredData.js";
import { useYScales } from "../../../hooks/useYScales.js";
const BarChartDisplay = ({
  series: barSeries,
  xScale
}) => {
  const yScale = useYScales();
  const { width: widthPlot, height: heightPlot } = useGridPlotSize();
  const { seriesMap, datapointsMap } = useTimeseriesFilteredData();
  return /* @__PURE__ */ jsx("figure", { "aria-label": BAR_CHART_DISPLAY_ARIA_LABEL, children: barSeries.map((series) => {
    const seriesMetadata = safeGet(seriesMap, series);
    return /* @__PURE__ */ jsx(
      BarCanvas,
      {
        data: series.datapoints,
        canvasWidth: widthPlot,
        canvasHeight: heightPlot,
        x: x(datapointsMap, xScale),
        y: y(datapointsMap, yScale),
        width: width(datapointsMap, xScale),
        height: height(datapointsMap, yScale),
        strokeWidth: 1,
        color: () => seriesMetadata.color.Default,
        opacity: () => DEFAULT_FILL_OPACITY
      },
      seriesMetadata.id
    );
  }) });
};
BarChartDisplay["displayName"] = "BarChartDisplay";
export {
  BarChartDisplay
};
