// @dynatrace/strato-components/tables/DataTable/features/GanttChart/DataTableGanttChartHeader.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DataTableGanttChartHeader_exports = {};
__export(DataTableGanttChartHeader_exports, {
  DataTableGanttChartHeader: () => DataTableGanttChartHeader
});
module.exports = __toCommonJS(DataTableGanttChartHeader_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_react = require("react");
var styles = __toESM(require("./DataTableGanttChartHeader.sty.js"));
var import_useMinMaxFromConfig = require("./hooks/useMinMaxFromConfig.js");
var import_useScaleAndFormatter = require("./hooks/useScaleAndFormatter.js");
var import_to_integration_tracks = require("./to-integration-tracks.js");
var import_AnnotationsNumericalDisplay = require("../../../../charts/core/components/annotations/components/displays/AnnotationsNumericalDisplay.js");
var import_AnnotationsTimeDisplay = require("../../../../charts/core/components/annotations/components/displays/AnnotationsTimeDisplay.js");
var import_AnnotationsScrollWrapper = require("../../../../charts/core/components/annotations/components/overlays/AnnotationsScrollWrapper.js");
var import_constants = require("../../../../charts/core/components/annotations/constants.js");
var import_TimeTableHeaderAxis = require("../../../../charts/core/components/axes/components/TimeTableHeaderAxis.js");
var import_ValueTableHeaderAxis = require("../../../../charts/core/components/axes/components/ValueTableHeaderAxis.js");
var import_ChartStateHandler = require("../../../../charts/core/components/chart-states/ChartStateHandler.js");
var import_coerceSizeValue = require("../../../../charts/core/utils/coerceSizeValue.js");
var import_useContainerSize = require("../../../../core/hooks/useContainerSize.js");
var import_Flex = require("../../../../layouts/flex/Flex.js");
var import_DataTableDefaultCell = require("../../components/DataTableDefaultCell.js");
var import_constants2 = require("../../constants.js");
var import_DataTableContext = require("../../contexts/DataTableContext.js");
const DataTableGanttChartHeader = ({
  columnDef
}) => {
  const sizeRef = (0, import_react.useRef)(null);
  const { width: axisWidth, height: axisHeight } = (0, import_useContainerSize.useContainerSize)(sizeRef);
  const config = columnDef.config || {};
  const { xAxisType = "numerical", annotationsHeader } = config;
  const { min, max } = (0, import_useMinMaxFromConfig.useMinMaxFromConfig)(columnDef.id);
  const table = (0, import_DataTableContext.useDataTable)();
  const isSubrowIndicatorColumn = table.getSubRowIndicatorColumnId() === columnDef.id;
  const rowDensity = table.getVisualVariantOptions().rowDensity;
  const leftHeaderPaddingValue = isSubrowIndicatorColumn ? import_constants2.DATA_TABLE_SUB_ROW_INDENT_GRID * table.getSubRowOpenDepth() : 0;
  const leftHeaderPadding = isSubrowIndicatorColumn ? `${leftHeaderPaddingValue}px` : "inherit";
  const subrowExpanderWidth = isSubrowIndicatorColumn ? import_constants2.DATA_TABLE_SUB_ROW_EXPANDER_ICON_SIZE + 2 * import_constants2.DATA_TABLE_DEFAULT_ICON_PADDING + import_constants2.DATA_TABLE_CELL_HORIZONTAL_PADDING_MAP[rowDensity] : 0;
  const effectiveAxisWidth = Math.max(
    0,
    axisWidth - leftHeaderPaddingValue - subrowExpanderWidth
  );
  const { scale, formatter } = (0, import_useScaleAndFormatter.useScaleAndFormatter)(
    effectiveAxisWidth,
    config,
    min,
    max
  );
  const {
    tooltip: annotationsTooltip,
    tooltipActions: annotationsTooltipActions,
    emptyState,
    errorState,
    loading,
    textOverflow = "truncate",
    truncateMode = "middle",
    tracks: annotationTracksConfig,
    height: annotationsHeightConfig
  } = annotationsHeader ?? {};
  const tooltipConfig = annotationsTooltip ? { template: annotationsTooltip } : {};
  const columnId = columnDef.id;
  const handleMarkerGroupHover = (0, import_react.useCallback)(
    (event) => {
      table.setGanttAnnotationsSelection(columnId, event);
    },
    [table, columnId]
  );
  const annotationTracks = annotationTracksConfig ? (0, import_to_integration_tracks.toIntegrationTracks)(annotationTracksConfig) : [];
  const shownTracks = annotationTracks.filter(({ hidden }) => !hidden);
  const tracksHeight = shownTracks.length * import_constants.TRACK_HEIGHT + import_constants.TRACK_SEPARATOR_STROKE_WIDTH;
  const configuredHeight = (0, import_coerceSizeValue.coerceSizeValue)(annotationsHeightConfig);
  const scrollWrapperHeight = configuredHeight ? `min(${tracksHeight}px, ${configuredHeight})` : tracksHeight;
  let ganttAxis = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ValueTableHeaderAxis.ValueTableHeaderAxis,
    {
      scale,
      formatter,
      containerHeight: axisHeight,
      containerWidth: effectiveAxisWidth,
      min,
      max
    }
  );
  let annotationsDisplay = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_AnnotationsNumericalDisplay.AnnotationsNumericalDisplay,
    {
      scale,
      tracks: annotationTracks,
      showLabels: false,
      labelsOffset: 0,
      width: effectiveAxisWidth,
      markerGroupHover: handleMarkerGroupHover,
      tooltipConfig,
      annotationsActions: annotationsTooltipActions,
      textOverflow,
      truncateMode,
      formatter
    }
  );
  if (xAxisType === "time") {
    ganttAxis = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_TimeTableHeaderAxis.TimeTableHeaderAxis,
      {
        scale,
        containerHeight: axisHeight,
        containerWidth: effectiveAxisWidth,
        min,
        max
      }
    );
    annotationsDisplay = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_AnnotationsTimeDisplay.AnnotationsTimeDisplay,
      {
        scale,
        tracks: annotationTracks,
        showLabels: false,
        labelsOffset: 0,
        width: effectiveAxisWidth,
        markerGroupHover: handleMarkerGroupHover,
        tooltipConfig,
        annotationsActions: annotationsTooltipActions,
        textOverflow,
        truncateMode
      }
    );
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_DataTableDefaultCell.DataTableDefaultCell,
    {
      ref: sizeRef,
      className: styles.ganttChartHeader,
      "data-sizeref": true,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
        import_Flex.Flex,
        {
          flexDirection: "column",
          minHeight: annotationsHeightConfig,
          justifyContent: "space-between",
          style: { paddingLeft: leftHeaderPadding },
          children: [
            annotationsHeader && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_ChartStateHandler.ChartStateHandler,
              {
                empty: (0, import_lodash_es.isEmpty)(annotationTracks),
                height: "100%",
                emptyState,
                errorState,
                isLoading: loading ?? false,
                style: { margin: "auto" },
                children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  import_AnnotationsScrollWrapper.AnnotationsScrollWrapper,
                  {
                    height: scrollWrapperHeight,
                    width: effectiveAxisWidth,
                    children: annotationsDisplay
                  }
                )
              }
            ),
            ganttAxis
          ]
        }
      )
    }
  );
};
