// @dynatrace/strato-components/tables/DataTable/features/GanttChart/DataTableGanttAnnotationsIndicators.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DataTableGanttAnnotationsIndicators_exports = {};
__export(DataTableGanttAnnotationsIndicators_exports, {
  DataTableGanttAnnotationIndicators: () => DataTableGanttAnnotationIndicators
});
module.exports = __toCommonJS(DataTableGanttAnnotationsIndicators_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var styles = __toESM(require("./DataTableGanttAnnotationsIndicators.sty.js"));
var import_constants = require("../../constants.js");
var import_useScaleAndFormatter = require("./hooks/useScaleAndFormatter.js");
var import_to_integration_tracks = require("./to-integration-tracks.js");
var import_AreaIndicator = require("../../../../charts/core/components/annotations/components/indicators/AreaIndicator.js");
var import_LineIndicator = require("../../../../charts/core/components/annotations/components/indicators/LineIndicator.js");
var import_get_shown_indicators = require("../../../../charts/core/components/annotations/components/indicators/utils/get-shown-indicators.js");
var import_get_unique_indicators_by_priority = require("../../../../charts/core/components/annotations/components/indicators/utils/get-unique-indicators-by-priority.js");
var import_get_marker_brightness_based_on_state = require("../../../../charts/core/components/annotations/utils/get-marker-brightness-based-on-state.js");
var import_is_range_marker = require("../../../../charts/core/components/annotations/utils/is-range-marker.js");
var import_serialize_marker_range = require("../../../../charts/core/components/annotations/utils/serialize-marker-range.js");
var import_BrightnessFilter = require("../../../../charts/core/components/BrightnessFilter.js");
var import_useCurrentTheme = require("../../../../core/hooks/useCurrentTheme.js");
var import_DataTableContext = require("../../contexts/DataTableContext.js");
function getBrightnessFilterId(indicator, trackIndex) {
  const { start, end } = indicator;
  return `gantt_brightness_${(0, import_serialize_marker_range.serializeMarkerRange)(
    {
      start: start.valueOf(),
      end: end ? end.valueOf() : void 0
    },
    trackIndex
  )}`;
}
function useGanttIndicators(config, columnId, range) {
  const table = (0, import_DataTableContext.useDataTable)();
  const { tracks: configuredTracks } = config.annotationsHeader ?? {};
  if (!configuredTracks?.length) {
    return null;
  }
  const selection = table.getState().ganttAnnotationsSelection?.[columnId] ?? {};
  const tracks = (0, import_to_integration_tracks.toIntegrationTracks)(configuredTracks);
  const shownIndicators = (0, import_get_shown_indicators.getShownIndicators)(tracks, selection);
  const uniqueIndicators = (0, import_get_unique_indicators_by_priority.getUniqueIndicatorsByPriority)(
    shownIndicators,
    range
  );
  const hasIndicators = uniqueIndicators.some((track) => track.length > 0);
  return hasIndicators ? uniqueIndicators : null;
}
function GanttIndicatorShape({
  indicator,
  trackIndex,
  scale,
  height,
  theme
}) {
  const { start, end, color, state } = indicator;
  const brightnessFilterId = getBrightnessFilterId(indicator, trackIndex);
  const slope = (0, import_get_marker_brightness_based_on_state.getMarkerBrightnessBasedOnState)(theme, state);
  const filter = `url(#${brightnessFilterId})`;
  const range = { start, end };
  const x = scale(start) ?? 0;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_BrightnessFilter.BrightnessFilter, { id: brightnessFilterId, slope }),
    (0, import_is_range_marker.isRangeMarker)(range) ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_AreaIndicator.AreaIndicator,
      {
        x,
        width: (scale(range.end) ?? 0) - x,
        height,
        color,
        filter
      }
    ) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_LineIndicator.LineIndicator, { x, height, color, filter })
  ] });
}
const DataTableGanttAnnotationIndicators = ({
  width,
  height,
  min,
  max,
  config,
  columnId,
  isLastPageRow
}) => {
  const theme = (0, import_useCurrentTheme.useCurrentTheme)();
  const { scale } = (0, import_useScaleAndFormatter.useScaleAndFormatter)(width, config, min, max);
  const indicators = useGanttIndicators(config, columnId, [min, max]);
  if (!indicators) {
    return null;
  }
  const fullRowHeight = height + (isLastPageRow ? 0 : import_constants.DATA_TABLE_ROW_BORDER);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "svg",
    {
      height: fullRowHeight,
      width,
      className: styles.annotationIndicators,
      children: indicators.map(
        (track, trackIndex) => track.map(([key, indicator]) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          GanttIndicatorShape,
          {
            indicator,
            trackIndex,
            scale,
            height: fullRowHeight,
            theme
          },
          key
        ))
      )
    }
  );
};
