// @dynatrace/strato-components/tables/DataTable/features/GanttChart/DataTableGanttChartGrid.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DataTableGanttChartGrid_exports = {};
__export(DataTableGanttChartGrid_exports, {
  DataTableGanttChartGrid: () => DataTableGanttChartGrid
});
module.exports = __toCommonJS(DataTableGanttChartGrid_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_scale = require("@visx/scale");
var import_react = require("react");
var import_borders = __toESM(require("@dynatrace/strato-design-tokens/borders"));
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
var styles = __toESM(require("./DataTableGanttChartGrid.sty.js"));
var import_useScaleAndFormatter = require("./hooks/useScaleAndFormatter.js");
var import_get_tick_value_as_number = require("../../../../charts/core/utils/get-tick-value-as-number.js");
var import_ticks = require("../../../../charts/core/utils/ticks.js");
const DataTableGanttChartGrid = ({
  width,
  height,
  min,
  max,
  config,
  isSubrowIndicatorColumn
}) => {
  const { xAxisType = "numerical" } = config;
  const { scale } = (0, import_useScaleAndFormatter.useScaleAndFormatter)(width, config, min, max);
  const numTicks = xAxisType === "time" ? (0, import_ticks.getGanttHorizontalTimeAxisNumTicks)(width) : (0, import_ticks.getHorizontalNumTicks)(width);
  const lines = (0, import_react.useMemo)(() => {
    const ticks = (0, import_scale.getTicks)(scale, numTicks);
    if (!isSubrowIndicatorColumn && (0, import_get_tick_value_as_number.getTickValueAsNumber)(ticks.at(0)?.valueOf()) === min) {
      ticks.shift();
    }
    if ((0, import_get_tick_value_as_number.getTickValueAsNumber)(ticks.at(-1)?.valueOf()) === max) {
      ticks.pop();
    }
    return ticks.map((rawTick) => {
      const tick = Math.max(scale(rawTick), 0.5);
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        "line",
        {
          x1: tick,
          x2: tick,
          y1: 0,
          y2: height,
          strokeWidth: import_borders.default.Width.Default,
          stroke: import_colors.default.Border.Neutral.Default
        },
        +rawTick
      );
    });
  }, [scale, numTicks, min, max, height]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "svg",
    {
      viewBox: `0 0 ${width} ${height}`,
      height,
      width,
      className: styles.ganttChartGrid,
      children: lines
    }
  );
};
