// @dynatrace/strato-components/tables/DataTable/features/GanttChart/GanttChart.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var GanttChart_exports = {};
__export(GanttChart_exports, {
  DataTableGanttChartFeature: () => DataTableGanttChartFeature,
  getMinMaxForGanttColumns: () => getMinMaxForGanttColumns,
  useGanttChart: () => useGanttChart
});
module.exports = __toCommonJS(GanttChart_exports);
var import_react = require("react");
var import_get_min_max_for_scale = require("../../../../charts/core/utils/get-min-max-for-scale.js");
var import_datatable_typeguards = require("../../hooks/useTable/datatable-typeguards.js");
var import_ColumnTypes = require("../ColumnTypes/ColumnTypes.js");
function isDataTableGanttData(obj) {
  return typeof obj === "object" && obj !== null && "start" in obj;
}
function getDataFromRow(row, columnDef) {
  if ((0, import_datatable_typeguards.isAccessorKeyColumnDef)(columnDef)) {
    return row[columnDef.accessorKey];
  }
  if ((0, import_datatable_typeguards.isAccessorFnColumnDef)(columnDef)) {
    return columnDef.accessorFn(row, 0);
  }
}
const adjustMinMaxWithAnnotationsHeader = (tracks, min, max) => {
  tracks.forEach((track) => {
    track.markers.forEach((marker) => {
      const start = marker.start;
      const end = marker.end ?? marker.start;
      const startValue = start instanceof Date ? start.valueOf() : start;
      const endValue = end instanceof Date ? end.valueOf() : end;
      min = Math.min(min, startValue);
      max = Math.max(max, endValue);
    });
  });
  return { min, max };
};
function getMinMaxForGanttColumns(data, columns, getSubRows) {
  const minMaxArray = columns.filter(
    (columnDef) => (0, import_datatable_typeguards.isDisplayColumnDef)(columnDef) && columnDef.columnType === "gantt"
  ).map((columnDef) => {
    const config = columnDef.meta?.config;
    let configMin = "data-min";
    let configMax = "data-max";
    let xAxisType = "numerical";
    let annotationsHeader;
    if ((0, import_ColumnTypes.isDatatableGanttColumnConfig)(config)) {
      configMin = config.min ?? configMin;
      configMax = config.max ?? configMax;
      xAxisType = config.xAxisType ?? xAxisType;
      annotationsHeader = config.annotationsHeader;
    }
    let ganttMin = Number.POSITIVE_INFINITY;
    let ganttMax = Number.NEGATIVE_INFINITY;
    const findGanttMinMax = (row) => {
      const ganttData = getDataFromRow(row, columnDef);
      const ganttSegments = Array.isArray(ganttData) ? ganttData : [ganttData];
      ganttSegments.forEach((segment) => {
        if (isDataTableGanttData(segment)) {
          const { start, end } = segment;
          ganttMin = Math.min(ganttMin, start);
          ganttMax = Math.max(ganttMax, end ?? start);
        }
      });
      const subRows = getSubRows?.(row, 0);
      subRows?.forEach((subRow) => findGanttMinMax(subRow));
    };
    for (const row of data) {
      findGanttMinMax(row);
    }
    if (annotationsHeader) {
      ({ min: ganttMin, max: ganttMax } = adjustMinMaxWithAnnotationsHeader(
        annotationsHeader.tracks,
        ganttMin,
        ganttMax
      ));
    }
    return [
      columnDef.id,
      xAxisType === "numerical" ? (0, import_get_min_max_for_scale.resolveMinMax)(
        { min: ganttMin, max: ganttMax },
        { min: configMin, max: configMax }
      ) : (0, import_get_min_max_for_scale.resolveTimeMinMax)(
        { min: ganttMin, max: ganttMax },
        { min: configMin, max: configMax }
      )
    ];
  });
  return Object.fromEntries(minMaxArray);
}
const DataTableGanttChartFeature = {
  createTable(table) {
    table.setGanttAnnotationsSelection = (columnId, selection) => {
      table.setState((state) => ({
        ...state,
        ganttAnnotationsSelection: {
          ...state.ganttAnnotationsSelection,
          [columnId]: {
            ...state.ganttAnnotationsSelection?.[columnId],
            ...selection
          }
        }
      }));
    };
  }
};
function useGanttChart(_props, options) {
  const ganttMinMaxValues = (0, import_react.useMemo)(
    () => getMinMaxForGanttColumns(
      options.data,
      options.columns,
      options.getSubRows
    ),
    [options.columns, options.data, options.getSubRows]
  );
  options.state = {
    ...options.state,
    ganttMinMaxValues
  };
}
