// @dynatrace/strato-geo/map/providers/imperative-handler.provider.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var imperative_handler_provider_exports = {};
__export(imperative_handler_provider_exports, {
  MapViewImperativeHandler: () => MapViewImperativeHandler
});
module.exports = __toCommonJS(imperative_handler_provider_exports);
var import_react_maplibre = require("@vis.gl/react-maplibre");
var import_react = require("react");
var import_constants = require("../constants.js");
var import_use_data_bounding_box = require("../hooks/use-data-bounding-box.js");
var import_use_initial_view_context = require("../hooks/use-initial-view-context.js");
var import_use_map_raw_data = require("../hooks/use-map-raw-data.js");
var import_store = require("../store/store.js");
var import_build_and_download_csv = require("../utils/build-and-download-csv.js");
const MapViewImperativeHandler = ({
  forwardedRef,
  containerRef,
  children
}) => {
  const layersData = (0, import_use_map_raw_data.useMapRawData)();
  const { current: map } = (0, import_react_maplibre.useMap)();
  const { longitude = 0, latitude = 0, zoom = 0 } = (0, import_use_initial_view_context.useInitialViewContext)();
  const boundingBox = (0, import_use_data_bounding_box.useDataBoundingBox)();
  const setZoomState = (0, import_store.useSetStateZoom)();
  const zoomOutEnabled = map && map?.getMinZoom() !== map?.getZoom();
  const zoomInEnabled = map && map?.getMaxZoom() !== map?.getZoom();
  const handleZoomToFit = () => {
    map?.fitBounds(boundingBox, import_constants.FIT_BOUNDS_OPTIONS);
    setZoomState({ zoomToFitEnabled: false, zoomToFitInProgress: true });
  };
  (0, import_react.useImperativeHandle)(forwardedRef, () => ({
    element: containerRef.current,
    downloadData: () => {
      (0, import_build_and_download_csv.buildAndDownloadCsv)(layersData);
    },
    zoomIn: () => zoomInEnabled && map?.zoomIn(),
    zoomOut: () => zoomOutEnabled && map?.zoomOut(),
    zoomToFit: () => handleZoomToFit(),
    reset: () => map?.flyTo({ center: [longitude, latitude], zoom })
  }));
  return children;
};
