// @dynatrace/strato-components/esm/charts/core/components/data-point/DataPointCanvas.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import Colors from "@dynatrace/strato-design-tokens/colors";
import { DATA_POINT_OUTER_RADIUS } from "./constants.js";
import { getStrokeColor, getOuterCircleSize } from "./utils/data-point.js";
import {
  DEFAULT_DATA_POINT_RADIUS,
  DEFAULT_HIGHLIGHTED_OPACITY
} from "../../constants.js";
import { RoundRectCanvas } from "../canvas/components/RoundRectCanvas.js";
import { SymbolCanvas } from "../canvas/components/SymbolCanvas.js";
const OuterCircleDataPoint = (props) => {
  const { points, closestPoint, valueAxis = "y", shape } = props;
  const { x, y, width, height } = getOuterCircleSize(points, valueAxis);
  const strokeColor = getStrokeColor(points, closestPoint);
  const isGroup = points.length > 1;
  return isGroup ? /* @__PURE__ */ jsx(
    RoundRectCanvas,
    {
      x,
      y,
      width,
      height,
      color: Colors.Background.Base.Default,
      opacity: DEFAULT_HIGHLIGHTED_OPACITY,
      borderRadius: 7,
      borderColor: strokeColor,
      borderWidth: 1
    }
  ) : /* @__PURE__ */ jsx(
    SymbolCanvas,
    {
      x: points[0]?.x,
      y: points[0]?.y,
      color: Colors.Background.Base.Default,
      opacity: DEFAULT_HIGHLIGHTED_OPACITY,
      r: DATA_POINT_OUTER_RADIUS,
      strokeColor,
      strokeWidth: 1,
      shape
    }
  );
};
const DataPointCanvas = (props) => {
  const { points, valueAxis = "y", shape } = props;
  const paintedPositions = /* @__PURE__ */ new Set();
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    points.length > 0 && /* @__PURE__ */ jsx(OuterCircleDataPoint, { ...props }),
    points.map((point, index) => {
      const key = Math.round(point[valueAxis]);
      if (paintedPositions.has(key) || point.id === "nothing") {
        return null;
      }
      paintedPositions.add(key);
      return /* @__PURE__ */ jsx(
        SymbolCanvas,
        {
          x: point.x,
          y: point.y,
          color: point.color.Default,
          r: DEFAULT_DATA_POINT_RADIUS,
          shape
        },
        index
      );
    })
  ] });
};
export {
  DataPointCanvas
};
