// @dynatrace/strato-geo/esm/map/components/BubbleLayer/BubbleLayer.js
import { jsx, jsxs } from "react/jsx-runtime";
import { Source } from "@vis.gl/react-maplibre";
import { _useBehavioralTrackingProps as useBehavioralTrackingProps } from "@dynatrace/strato-components/core";
import { BubbleCircleLayer } from "./BubbleCircleLayer.js";
import { BubbleLayerTooltip } from "./BubbleLayerTooltip.js";
import { BubbleOutlineLayer } from "./BubbleOutlineLayer.js";
import { SizeInterpolationContext } from "./contexts/size-interpolation.context.js";
import { buildRadiusScale } from "./utils/build-radius-scale.js";
import { parseBubbleDataToGeoJSON } from "./utils/parse-bubble-data-to-geo-json.js";
import { DEFAULT_BUBBLE_COLOR, DEFAULT_RADIUS } from "../../constants.js";
import { useLayerColoringStrategy } from "../../hooks/use-layer-coloring-strategy.js";
import { useLegendInteractionHighlighting } from "../../hooks/use-legend-interaction-highlighting.js";
import { useLegendInteractionVisibility } from "../../hooks/use-legend-interaction-visibility.js";
import { useResolveLocationColor } from "../../hooks/use-resolve-color.js";
import { useTooltipTemplate } from "../../hooks/use-tooltip-template.js";
import { BubbleLayerTooltip as BubbleLayerTooltipSlot } from "../../slots/BubbleLayerTooltip.js";
import {
  isScaleRadius
} from "../../types/bubble-layer.js";
import { isBubbleTooltipTemplate } from "../../utils/tooltip-type-guards.js";
import { TooltipWrapper } from "../TooltipWrapper.js";
const BubbleLayer = (props) => {
  const [behavioralTrackingProps] = useBehavioralTrackingProps(props);
  const {
    data,
    layerId,
    radius = DEFAULT_RADIUS,
    children,
    sizeInterpolation = "fixed"
  } = props;
  const colorParser = useLayerColoringStrategy();
  const bubbleColor = useResolveLocationColor(
    DEFAULT_BUBBLE_COLOR,
    props,
    colorParser
  );
  const sourceId = `source-${layerId}`;
  useLegendInteractionHighlighting(props, sourceId);
  useLegendInteractionVisibility(props, layerId);
  let radiusScale = (value) => value;
  if (isScaleRadius(props)) {
    const { radiusRange = [10, 100], radius: radius2, scale = "linear" } = props;
    radiusScale = buildRadiusScale(data, radius2, scale, radiusRange);
  }
  const dataGeoJson = parseBubbleDataToGeoJSON(
    data,
    bubbleColor,
    radius,
    radiusScale
  );
  const {
    handler: tooltipTemplate,
    hidden: tooltipHidden,
    symbolAlignment
  } = useTooltipTemplate(children, BubbleLayerTooltipSlot, {
    extractProps: true
  });
  return /* @__PURE__ */ jsx(SizeInterpolationContext.Provider, { value: sizeInterpolation, children: /* @__PURE__ */ jsxs(Source, { id: sourceId, type: "geojson", data: dataGeoJson, generateId: true, children: [
    /* @__PURE__ */ jsx(
      BubbleCircleLayer,
      {
        id: layerId,
        source: sourceId,
        ...behavioralTrackingProps
      }
    ),
    /* @__PURE__ */ jsx(BubbleOutlineLayer, { beforeId: layerId, source: sourceId }),
    /* @__PURE__ */ jsx(TooltipWrapper, { layerId, children: /* @__PURE__ */ jsx(
      BubbleLayerTooltip,
      {
        tooltipTemplate: isBubbleTooltipTemplate(tooltipTemplate) ? tooltipTemplate : void 0,
        hidden: tooltipHidden,
        symbolAlignment
      }
    ) })
  ] }) });
};
BubbleLayer["displayName"] = "BubbleLayer";
export {
  BubbleLayer
};
