// @dynatrace/strato-components/charts/core/components/chart-grid/ChartGridRoot.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ChartGridRoot_exports = {};
__export(ChartGridRoot_exports, {
  ChartGrid: () => ChartGrid,
  ChartGridRoot: () => ChartGridRoot
});
module.exports = __toCommonJS(ChartGridRoot_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_ChartGridArea = require("./ChartGridArea.js");
var import_ChartGridPlot = require("./ChartGridPlot.js");
var import_GridPlotSizeProvider = require("./providers/GridPlotSizeProvider.js");
var import_useMergeRefs = require("../../../../core/hooks/useMergeRefs.js");
var import_Grid = require("../../../../layouts/grid/Grid.js");
const ChartGridRoot = /* @__PURE__ */ (0, import_react.forwardRef)(
  ({
    width,
    height,
    gridTemplateAreas,
    gridTemplateColumns,
    gridTemplateRows,
    children,
    ...props
  }, forwardedRef) => {
    const chartRef = (0, import_react.useRef)(null);
    const mergedRef = (0, import_useMergeRefs.useMergeRefs)([forwardedRef, chartRef]);
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Grid.Grid,
      {
        ref: mergedRef,
        width,
        height,
        gridTemplateAreas,
        gridTemplateColumns,
        gridTemplateRows,
        minHeight: "min-content",
        gap: 0,
        ...props,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_GridPlotSizeProvider.GridPlotSizeProvider, { children })
      }
    );
  }
);
ChartGridRoot.displayName = "ChartGridRoot";
const ChartGrid = Object.assign(ChartGridRoot, {
  PlotArea: import_ChartGridPlot.ChartGridPlot,
  GridArea: import_ChartGridArea.ChartGridArea
});
