// @dynatrace/strato-components/content/empty-state/EmptyState.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var EmptyState_exports = {};
__export(EmptyState_exports, {
  EmptyState: () => EmptyState
});
module.exports = __toCommonJS(EmptyState_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_empty_state_utils = require("./empty-state-utils.js");
var import_EmptyState_css = require("./EmptyState.sty.js");
var import_EmptyStateWrapper = require("./EmptyStateWrapper.js");
var import_Actions = require("./slot-components/Actions.js");
var import_Details = require("./slot-components/Details.js");
var import_Footer = require("./slot-components/Footer.js");
var import_Title = require("./slot-components/Title.js");
var import_Visual = require("./slot-components/Visual.js");
var import_VisualPreset = require("./slot-components/VisualPreset.js");
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useSafeForwardProps = require("../../core/hooks/useSafeForwardProps.js");
var import_useSlots = require("../../core/hooks/useSlots.js");
var import_Divider = require("../../layouts/divider/Divider.js");
var import_Flex = require("../../layouts/flex/Flex.js");
var import_Heading = require("../../typography/heading/Heading.js");
const COMPONENT_NAME = "EmptyState";
const BaseEmptyState = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    size = "default",
    children,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const ariaLabelingProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(restProps);
  const forwardProps = (0, import_useSafeForwardProps.useSafeForwardProps)(restProps, ariaLabelingProps);
  const [slots] = (0, import_useSlots.useSlots)(children, {
    Actions: import_Actions.Actions,
    Details: import_Details.Details,
    Footer: import_Footer.Footer,
    Title: import_Title.Title,
    Visual: import_Visual.Visual,
    VisualPreset: import_Visual.Visual
  });
  const visualSlot = slots.Visual ? slots.Visual : null;
  const detailsSlot = slots.Details ? slots.Details : null;
  const actionsSlot = slots.Actions ? slots.Actions : null;
  const footerSlot = slots.Footer ? slots.Footer : null;
  const titleSlot = slots.Title ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Heading.Heading,
    {
      ...(0, import_empty_state_utils.getHeadingVariant)(size, slots.Title.props.as),
      className: (0, import_clsx.default)(import_EmptyState_css.titleSlotCss),
      children: slots.Title
    }
  ) : null;
  const hasDivider = (slots.Actions || slots.Details || slots.Title || slots.Visual) && slots.Footer;
  const divider = hasDivider ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Divider.Divider, {}) : null;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_EmptyStateWrapper.EmptyStateWrapper,
    {
      size,
      ref: forwardedRef,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...ariaLabelingProps,
      ...behavioralTrackingProps,
      ...forwardProps,
      "data-dt-component": COMPONENT_NAME,
      children: [
        visualSlot,
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
          import_Flex.Flex,
          {
            className: (0, import_clsx.default)(import_EmptyState_css.contentContainerCss),
            maxWidth: 600,
            flexDirection: "column",
            gap: 0,
            children: [
              titleSlot,
              detailsSlot,
              actionsSlot,
              divider,
              footerSlot
            ]
          }
        )
      ]
    }
  );
});
BaseEmptyState.displayName = COMPONENT_NAME;
const EmptyState = Object.assign(BaseEmptyState, {
  Actions: import_Actions.Actions,
  Details: import_Details.Details,
  Footer: import_Footer.Footer,
  Title: import_Title.Title,
  Visual: import_Visual.Visual,
  VisualPreset: import_VisualPreset.VisualPreset
});
