// @dynatrace/strato-components/content/empty-state/EmptyState.sty.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var EmptyState_css_exports = {};
__export(EmptyState_css_exports, {
  actionsSlotCss: () => actionsSlotCss,
  contentContainerCss: () => contentContainerCss,
  detailsSlotCss: () => detailsSlotCss,
  emptyStateContainerCss: () => emptyStateContainerCss,
  footerSlotCss: () => footerSlotCss,
  titleSlotCss: () => titleSlotCss,
  visualSlotCss: () => visualSlotCss
});
module.exports = __toCommonJS(EmptyState_css_exports);
var import_EmptyState_css_ts_vanilla = require("./EmptyState.css");
var import_createRuntimeFn = require("@vanilla-extract/recipes/createRuntimeFn");
var actionsSlotCss = "_1os652u9-3-6-0";
var contentContainerCss = "_1os652u6-3-6-0";
var detailsSlotCss = "_1os652u8-3-6-0";
var emptyStateContainerCss = (0, import_createRuntimeFn.createRuntimeFn)({ defaultClassName: "_1os652u0-3-6-0", variantClassNames: { size: { small: "_1os652u1-3-6-0", "default": "_1os652u2-3-6-0", large: "_1os652u3-3-6-0", largeResponsive: "_1os652u4-3-6-0" } }, defaultVariants: {}, compoundVariants: [] });
var footerSlotCss = "_1os652ua-3-6-0";
var titleSlotCss = "_1os652u7-3-6-0";
var visualSlotCss = "_1os652u5-3-6-0";
