// @dynatrace/strato-components/core/contexts/ScrollContext.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ScrollContext_exports = {};
__export(ScrollContext_exports, {
  ScrollContext: () => ScrollContext
});
module.exports = __toCommonJS(ScrollContext_exports);
var import_react = require("react");
const ScrollContext = (0, import_react.createContext)({
  /** the x and y position of the container ref */
  scrollPosition: {
    x: 0,
    y: 0
  },
  /** while container is scrolling - isScrolling is true */
  isScrolling: false
});
ScrollContext["displayName"] = "ScrollContext";
