// @dynatrace/strato-components/forms/spin-buttons/text-spin-button/hooks/useValueNavigation.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useValueNavigation_exports = {};
__export(useValueNavigation_exports, {
  useValueNavigation: () => useValueNavigation
});
module.exports = __toCommonJS(useValueNavigation_exports);
var import_react = require("react");
var import_helper = require("../../helper.js");
var import_constants = require("../constants.js");
const useValueNavigation = (props) => {
  const {
    options,
    onArrowUp = (idx, e) => {
      e.preventDefault();
      return idx - import_constants.TEXT_SPIN_BUTTON_DEFAULTS.SMALL_STEP;
    },
    onArrowDown = (idx, e) => {
      e.preventDefault();
      return idx + import_constants.TEXT_SPIN_BUTTON_DEFAULTS.SMALL_STEP;
    },
    onPageUp = (idx, e) => {
      e.preventDefault();
      return idx - import_constants.TEXT_SPIN_BUTTON_DEFAULTS.BIG_STEP;
    },
    onPageDown = (idx, e) => {
      e.preventDefault();
      return idx + import_constants.TEXT_SPIN_BUTTON_DEFAULTS.BIG_STEP;
    },
    onHome = (_, e) => {
      e.preventDefault();
      return 0;
    },
    onEnd = (_, e) => {
      e.preventDefault();
      return options.length - 1;
    },
    disabled,
    readOnly,
    defaultValue = import_constants.TEXT_SPIN_BUTTON_DEFAULTS.VALUE,
    value: valueProp,
    onChange,
    isCyclic
  } = props;
  const [value, setValue] = (0, import_react.useState)(valueProp ?? defaultValue);
  const handleValueNavigation = (0, import_react.useCallback)(
    (transform, event) => {
      const sanitized = (idx) => {
        const maxIndex = options.length - 1;
        return (0, import_helper.constrainValue)(idx, 0, maxIndex, isCyclic) ?? 0;
      };
      const curIndex = value === null ? 0 : options.indexOf(value || valueProp || "");
      const newIndex = sanitized(transform(curIndex, event));
      const newValue = options[newIndex];
      if (value !== newValue) {
        setValue(newValue);
        onChange?.(newValue);
      }
    },
    [value, options, valueProp, onChange, isCyclic]
  );
  const lookupTable = (0, import_react.useMemo)(
    () => options.map((v) => v[0].toLowerCase()),
    [options]
  );
  const searchValue = (0, import_react.useCallback)(
    (key, fireOnChange = false) => {
      const idx = lookupTable.indexOf(key.toLowerCase());
      if (idx < 0) {
        return void 0;
      }
      const newValue = options[idx];
      if (value !== newValue) {
        setValue(newValue);
      }
      fireOnChange && onChange?.(newValue);
      return newValue;
    },
    [lookupTable, options, value, onChange]
  );
  const clearValue = (0, import_react.useCallback)(
    (fireOnChange = false) => {
      setValue(null);
      fireOnChange && onChange?.(null);
    },
    [onChange]
  );
  const onKeyDownHandler = (0, import_react.useCallback)(
    (e) => {
      if (disabled || readOnly) {
        return;
      }
      const navigationHandlers = {
        ArrowUp: onArrowUp,
        ArrowDown: onArrowDown,
        PageUp: onPageUp,
        PageDown: onPageDown,
        Home: onHome,
        End: onEnd
      };
      const navHandler = navigationHandlers[e.key];
      if (navHandler) {
        if (!e.shiftKey && e.key !== "Shift") {
          handleValueNavigation(navHandler, e);
        }
        return;
      }
      switch (e.key) {
        case "Backspace":
        case "Delete":
          if (value !== null) {
            setValue(null);
            onChange?.(null);
          }
          e.preventDefault();
          break;
        default:
          if (!e.metaKey && !e.ctrlKey && e.key.match(/^\w$/)) {
            const newValue = searchValue(e.key);
            if (newValue !== void 0 && newValue !== value) {
              onChange?.(newValue);
            }
            e.preventDefault();
          }
          break;
      }
    },
    [
      disabled,
      handleValueNavigation,
      onArrowDown,
      onArrowUp,
      onChange,
      onEnd,
      onHome,
      onPageDown,
      onPageUp,
      readOnly,
      searchValue,
      setValue,
      value
    ]
  );
  (0, import_react.useEffect)(() => {
    if (valueProp !== void 0) {
      setValue(valueProp);
    }
  }, [setValue, valueProp]);
  return { value, onKeyDownHandler, searchValue, clearValue };
};
