// @dynatrace/strato-components/esm/core/dql/dql-common.js
const dqlCommands = [
  "aggregate",
  "append",
  "data",
  "dedup",
  "describe",
  "expand",
  "experimentalFieldsSummary",
  "experimentalLoad",
  "experimentalSave",
  "experimentalTimeChart",
  "failIf",
  "fetch",
  "fields",
  "fieldsAdd",
  "fieldsFlatten",
  "fieldsKeep",
  "fieldsRemove",
  "fieldsRename",
  "fieldsSnapshot",
  "fieldsSummary",
  "filter",
  "filterOut",
  "group",
  "internalTimeseriesBins",
  "join",
  "joinNested",
  "limit",
  "load",
  "lookup",
  "makeTimeseries",
  "parse",
  "search",
  "smartscapeEdges",
  "smartscapeNodes",
  "sort",
  "summarize",
  "timeseries",
  "traverse"
];
const sortedDqlCommands = dqlCommands.toSorted(
  (a, b) => b.length - a.length
);
const parameterNamePattern = /\w+\s*(?=:)/;
const booleanLiteralPattern = /\b(?:true|false)\b/i;
const nullLiteralPattern = /\bnull\b/i;
const dqlDurationUnits = [
  "minutes",
  "minute",
  "hours",
  "hour",
  "mins",
  "secs",
  "hrs",
  "min",
  "sec",
  "ms",
  "ns",
  "us",
  "hr",
  "h",
  "s",
  "m"
];
const dqlCalendarDurationUnits = [
  "quarters",
  "quarter",
  "months",
  "weeks",
  "years",
  "month",
  "qtrs",
  "week",
  "year",
  "days",
  "day",
  "qtr",
  "yrs",
  "yr",
  "M",
  "w",
  "d",
  "q",
  "y"
];
const combinedDurationUnits = [
  ...dqlDurationUnits,
  ...dqlCalendarDurationUnits
];
const dqlUnitSuffixPattern = `(?:${combinedDurationUnits.join("|")})`;
export {
  booleanLiteralPattern,
  combinedDurationUnits,
  dqlCommands,
  dqlUnitSuffixPattern,
  nullLiteralPattern,
  parameterNamePattern,
  sortedDqlCommands
};
