// @dynatrace/strato-components/charts/categorical-bar/contexts/CategoricalBarFilteredData.context.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CategoricalBarFilteredData_context_exports = {};
__export(CategoricalBarFilteredData_context_exports, {
  CategoricalBarFilteredDataContext: () => CategoricalBarFilteredDataContext
});
module.exports = __toCommonJS(CategoricalBarFilteredData_context_exports);
var import_react = require("react");
const CategoricalBarFilteredDataContext = (0, import_react.createContext)({
  filteredCategories: [],
  filteredDimensions: [],
  isSingleDimension: false,
  minMaxPerUnit: /* @__PURE__ */ new Map(),
  categoryMap: /* @__PURE__ */ new WeakMap(),
  dimensionMap: /* @__PURE__ */ new WeakMap(),
  unit: void 0
});
