// @dynatrace/strato-components/esm/navigation/tabs/RenderTab.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import {
  useRef
} from "react";
import {
  disabledTabWrapperCSS,
  tabButtonCSS,
  tabContentWrapperCSS
} from "./RenderTab.sty.js";
import { useFocusRing } from "../../core/styles/useFocusRing.js";
import { Text } from "../../typography/text/Text.js";
const RenderTab = (props) => {
  const {
    isDisabled: propsDisabled = false,
    isSelected = false,
    tabItem,
    ariaControls,
    onActivate,
    onKeyDown
  } = props;
  const ref = useRef(null);
  const { key, props: tabProps } = tabItem;
  const {
    title,
    disabled,
    children,
    prefixIcon,
    id = key ?? void 0,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    keepMounted,
    ...restProps
  } = tabProps;
  const isDisabled = propsDisabled || (disabled ?? false);
  const { focusProps, focusClassName } = useFocusRing({
    color: !isDisabled && isSelected ? "primary" : "neutral",
    disabled: isDisabled
  });
  return /* @__PURE__ */ jsx("div", { className: clsx({ [disabledTabWrapperCSS]: isDisabled }), children: /* @__PURE__ */ jsx(
    "button",
    {
      "data-testid": dataTestId,
      ...restProps,
      ref,
      type: "button",
      "data-key": key,
      role: "tab",
      id,
      disabled: isDisabled,
      "aria-selected": isSelected,
      "aria-controls": ariaControls,
      ...!isDisabled && !propsDisabled && {
        onClick: onActivate,
        onKeyDown,
        onBlur: focusProps.onBlur,
        onFocus: focusProps.onFocus
      },
      className: clsx(
        tabButtonCSS({
          disabled: isDisabled,
          selected: isSelected
        }),
        focusClassName,
        consumerClassName
      ),
      style: consumerStyle,
      children: /* @__PURE__ */ jsxs("div", { className: tabContentWrapperCSS, children: [
        prefixIcon !== void 0 && prefixIcon,
        /* @__PURE__ */ jsx(Text, { className: "text-content", as: "div", textStyle: "base", children: title })
      ] })
    }
  ) });
};
export {
  RenderTab
};
