// @dynatrace/strato-components/esm/navigation/tabs/TabPanel.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import { useLayoutEffect, useRef, useState } from "react";
import { tabPanelCSS, tabPanelOverflowBehaviorCSS } from "./TabPanel.sty.js";
import { useFocusRing } from "../../core/styles/useFocusRing.js";
import { getFirstFocusableChild as _getFirstFocusableChild } from "../../core/utils/focus-management/get-first-focusable-child.js";
import { mergeProps as _mergeProps } from "../../core/utils/merge-props.js";
const TabPanel = (props) => {
  const {
    children,
    disabled,
    panelOverflow,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    ...restProps
  } = props;
  const ref = useRef(null);
  const { focusProps, focusClassName } = useFocusRing({ disabled });
  const [hasFocusableChildren, setHasFocusableChildren] = useState(false);
  useLayoutEffect(() => {
    if (ref.current) {
      setHasFocusableChildren(_getFirstFocusableChild(ref.current) !== null);
    }
  }, [children]);
  return /* @__PURE__ */ jsx(
    "div",
    {
      "data-testid": dataTestId,
      className: clsx(
        consumerClassName,
        focusClassName,
        tabPanelCSS,
        tabPanelOverflowBehaviorCSS[panelOverflow]
      ),
      style: consumerStyle,
      ref,
      ..._mergeProps(focusProps, restProps),
      tabIndex: hasFocusableChildren ? -1 : 0,
      role: "tabpanel",
      children: !disabled && children
    }
  );
};
TabPanel["displayName"] = "TabPanel";
export {
  TabPanel
};
