// @dynatrace/strato-components/esm/navigation/tabs/Tabs.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import {
  Children,
  cloneElement,
  forwardRef,
  useCallback,
  useLayoutEffect,
  useMemo,
  useRef,
  useState
} from "react";
import { useIntl } from "react-intl";
import { isElement } from "react-is";
import useResizeObserver from "use-resize-observer";
import { ChevronLeftIcon, ChevronRightIcon } from "@dynatrace/strato-icons";
import { RenderTab } from "./RenderTab.js";
import { Tab } from "./Tab.js";
import { TabPanel } from "./TabPanel.js";
import {
  tabsArrowButtonCSS,
  tabsArrowWrapperCSS,
  tabsCSS,
  tabsWrapperCSS,
  tabsWrapperOverflowCSS
} from "./Tabs.sty.js";
import { Button } from "../../buttons/button/Button.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useControllableState } from "../../core/hooks/useControllableState.js";
import { useId } from "../../core/hooks/useId.js";
import { useMergeRefs } from "../../core/hooks/useMergeRefs.js";
import { useRovingFocus } from "../../core/hooks/useRovingFocus.js";
import { isDev } from "../../core/utils/is-node-environment.js";
import { isSlot } from "../../core/utils/is-slot.js";
const COMPONENT_NAME = "Tab";
const Tabs = /* @__PURE__ */ forwardRef(
  (props, forwardedRef) => {
    const {
      disabled: disabledTabsProp = false,
      selectedIndex: selectedIndexProp,
      defaultIndex: defaultIndexProp,
      onChange: onChangeProp,
      children,
      panelOverflow = "hidden",
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...remainingProps
    } = props;
    const intl = useIntl();
    const generatedId = useId();
    const tabsUniqueId = props.id ?? generatedId;
    const tabsListId = useId();
    const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
    const childrenArray = useMemo(() => {
      return Children.map(children, (child, i) => {
        if (isSlot(child, Tab)) {
          return cloneElement(child, {
            key: `${tabsUniqueId}-tab-${i}`
          });
        } else if (isElement(child) && isDev()) {
          console.warn("Only Tab is a valid child of the Tabs component.");
        }
      }) || [];
    }, [children, tabsUniqueId]);
    const tabsInfo = useMemo(() => {
      return childrenArray.reduce(
        (aggr, item) => {
          if (item.props.id && item.key) {
            aggr.consumerIds[item.key] = item.props.id;
          }
          if (item.props.disabled && item.key !== null) {
            aggr.disabled.add(item.key);
          }
          if (item.props.keepMounted && item.key !== null) {
            aggr.keepMounted.add(item.key);
          }
          return aggr;
        },
        {
          disabled: /* @__PURE__ */ new Set(),
          keepMounted: /* @__PURE__ */ new Set(),
          consumerIds: {}
        }
      );
    }, [childrenArray]);
    const [selectedIndex, setSelectedIndex] = useControllableState({
      defaultValue: defaultIndexProp ?? 0,
      value: selectedIndexProp,
      onChange: onChangeProp
    });
    const ref = useRef(null);
    useLayoutEffect(() => {
      const tabList = ref.current;
      if (!tabList) {
        return;
      }
      const selectedTab = tabList.querySelector(
        '[aria-selected="true"]'
      );
      if (!selectedTab) {
        return;
      }
      const tabListRect = tabList.getBoundingClientRect();
      const tabRect = selectedTab.getBoundingClientRect();
      const leftArrow = tabList.previousElementSibling;
      const rightArrow = tabList.nextElementSibling;
      const leftOffset = leftArrow?.offsetWidth ?? 0;
      const rightOffset = rightArrow?.offsetWidth ?? 0;
      const ADJACENT_TAB_PEEK_WIDTH_PX = 24;
      const isHiddenLeft = tabRect.left < tabListRect.left + leftOffset;
      const isHiddenRight = tabRect.right > tabListRect.right - rightOffset;
      if (isHiddenLeft) {
        const targetLeftPosition = tabListRect.left + leftOffset + ADJACENT_TAB_PEEK_WIDTH_PX;
        const scrollAmount = targetLeftPosition - tabRect.left;
        tabList.scrollLeft -= scrollAmount;
      } else if (isHiddenRight) {
        const targetRightPosition = tabListRect.right - rightOffset - ADJACENT_TAB_PEEK_WIDTH_PX;
        const scrollAmount = tabRect.right - targetRightPosition;
        tabList.scrollLeft += scrollAmount;
      }
    }, [selectedIndex]);
    const { containerRef } = useRovingFocus({
      direction: "both",
      focusableChildrenSelector: '[type="button"],[role="tab"]',
      initialIndex: selectedIndex
    });
    const rootRef = useMergeRefs([ref, containerRef]);
    const [isOverflowingLeft, setIsOverflowingLeft] = useState(false);
    const [isOverflowingRight, setIsOverflowingRight] = useState(false);
    const slideTabsLeft = () => {
      ref.current?.scrollBy({
        left: -ref.current.offsetWidth / 2,
        behavior: "smooth"
      });
    };
    const slideTabsRight = () => {
      ref.current?.scrollBy({
        left: ref.current.offsetWidth / 2,
        behavior: "smooth"
      });
    };
    const updateOverflow = useCallback(() => {
      if (ref.current) {
        const tabsWidth = ref.current.offsetWidth;
        const tabsScrollWidth = ref.current.scrollWidth;
        const scrollDiff = tabsScrollWidth - tabsWidth;
        const tabsScrollLeft = Math.ceil(ref.current.scrollLeft);
        setIsOverflowingRight(
          tabsScrollWidth > tabsWidth && tabsScrollLeft < scrollDiff
        );
        setIsOverflowingLeft(
          tabsScrollWidth > tabsWidth && tabsScrollLeft !== 0
        );
      }
    }, []);
    const { width: containerWidth, height: containerHeight } = useResizeObserver({
      ref: ref.current
    });
    useLayoutEffect(() => {
      updateOverflow();
    }, [containerWidth, containerHeight, updateOverflow]);
    const getSelectedTabContent = (key, childrenArray2) => {
      const tab = childrenArray2.find((child) => child.key === key);
      if (tab?.props?.children) {
        return tab.props.children;
      }
      return null;
    };
    const getTabPanelId = (key, options) => {
      if (!options || options.keepMounted) {
        return key !== null ? `${key}-tabPanel` : void 0;
      }
      return key !== null && options.isSelected ? `${key}-tabPanel` : void 0;
    };
    const MemoizedSelectedTabPanel = useMemo(
      () => childrenArray.map((child, index) => {
        const isSelected = index === selectedIndex;
        const childKey = child.key;
        if (childKey !== null && (tabsInfo.keepMounted.has(childKey) || isSelected)) {
          return /* @__PURE__ */ jsx(
            TabPanel,
            {
              style: {
                display: !isSelected ? "none" : ""
              },
              id: getTabPanelId(childKey),
              disabled: disabledTabsProp || tabsInfo.disabled.has(childKey),
              panelOverflow,
              "aria-labelledby": tabsInfo.consumerIds[childKey] || childKey,
              children: getSelectedTabContent(childKey, childrenArray)
            },
            childKey
          );
        }
        return null;
      }),
      [childrenArray, disabledTabsProp, panelOverflow, selectedIndex, tabsInfo]
    );
    const handleKeyDown = useCallback(
      (event) => {
        const selectNextTab = () => {
          setSelectedIndex(
            (curr) => curr + 1 > childrenArray.length - 1 ? 0 : curr + 1
          );
        };
        const selectPrevTab = () => {
          setSelectedIndex(
            (curr) => curr - 1 < 0 ? childrenArray.length - 1 : curr - 1
          );
        };
        if (["ArrowRight", "ArrowDown"].includes(event.code)) {
          selectNextTab();
        }
        if (["ArrowLeft", "ArrowUp"].includes(event.code)) {
          selectPrevTab();
        }
      },
      [setSelectedIndex, childrenArray.length]
    );
    const MemoizedTabsList = useMemo(() => {
      return childrenArray.map((item, index) => /* @__PURE__ */ jsx(
        RenderTab,
        {
          tabItem: item,
          isSelected: selectedIndex === index && !item.props.disabled && !disabledTabsProp,
          isDisabled: item.props.disabled || disabledTabsProp,
          onActivate: () => {
            setSelectedIndex(index);
          },
          onKeyDown: handleKeyDown,
          ariaControls: getTabPanelId(item.key, {
            keepMounted: item.props.keepMounted,
            isSelected: selectedIndex === index && !item.props.disabled && !disabledTabsProp
          })
        },
        item.key
      ));
    }, [
      childrenArray,
      selectedIndex,
      disabledTabsProp,
      setSelectedIndex,
      handleKeyDown
    ]);
    return /* @__PURE__ */ jsxs(
      "div",
      {
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        ref: forwardedRef,
        className: clsx(tabsCSS, consumerClassName),
        style: consumerStyle,
        ...restProps,
        ...behavioralTrackingProps,
        "data-dt-component": COMPONENT_NAME,
        children: [
          /* @__PURE__ */ jsxs("div", { className: tabsArrowWrapperCSS, children: [
            isOverflowingLeft && /* @__PURE__ */ jsx(
              Button,
              {
                className: tabsArrowButtonCSS,
                style: { left: 0 },
                variant: "default",
                onClick: slideTabsLeft,
                "aria-label": intl.formatMessage({
                  defaultMessage: "scroll tabs left",
                  id: "Cv2AP7VKybeuopCP",
                  description: "Aria label to scroll the tabs to the left"
                }),
                "aria-controls": tabsListId,
                children: /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(ChevronLeftIcon, {}) })
              }
            ),
            /* @__PURE__ */ jsx(
              "div",
              {
                className: clsx(tabsWrapperCSS, {
                  [tabsWrapperOverflowCSS.left]: isOverflowingLeft && !isOverflowingRight,
                  [tabsWrapperOverflowCSS.right]: !isOverflowingLeft && isOverflowingRight,
                  [tabsWrapperOverflowCSS.both]: isOverflowingLeft && isOverflowingRight
                }),
                role: "tablist",
                "aria-orientation": "horizontal",
                id: tabsListId,
                ref: rootRef,
                tabIndex: -1,
                onScroll: updateOverflow,
                children: MemoizedTabsList
              }
            ),
            isOverflowingRight && /* @__PURE__ */ jsx(
              Button,
              {
                className: tabsArrowButtonCSS,
                style: { right: 0 },
                onClick: slideTabsRight,
                "aria-label": intl.formatMessage({
                  defaultMessage: "scroll tabs right",
                  id: "2YBMATuaXx9/4zBP",
                  description: "Aria label to scroll the tabs to the right"
                }),
                "aria-controls": tabsListId,
                children: /* @__PURE__ */ jsx(Button.Suffix, { children: /* @__PURE__ */ jsx(ChevronRightIcon, {}) })
              }
            )
          ] }),
          MemoizedSelectedTabPanel
        ]
      }
    );
  }
);
Tabs.displayName = "Tabs";
export {
  Tabs
};
