// @dynatrace/strato-components/filters/segment-selector/hooks/useSegments.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useSegments_exports = {};
__export(useSegments_exports, {
  useSegments: () => useSegments
});
module.exports = __toCommonJS(useSegments_exports);
var import_react_query = require("@tanstack/react-query");
var import_react = require("react");
var import_segments_presets = require("@dynatrace-sdk/segments-presets");
var import_selector = require("../../../core/store/selector.js");
var import_is_node_environment = require("../../../core/utils/is-node-environment.js");
var import_queryHelper = require("../../helper/queryHelper.js");
var import_SegmentsCache = require("../contexts/SegmentsCache/SegmentsCache.js");
var import_SegmentsContext = require("../contexts/SegmentsStore/SegmentsContext.js");
var import_SegmentsUIDispatchContext = require("../contexts/SegmentsUI/SegmentsUIDispatchContext.js");
var import_segmentValidationHelper = require("../contexts/SegmentsUI/segmentValidationHelper.js");
var import_isSelectedPreset = require("../parts/segmentsPresets/isSelectedPreset.js");
function updateCache(segments, old) {
  if (!old) {
    return { recentPresets: [], pinnedPresets: [] };
  }
  const recentPresets = old.recentPresets;
  const recentPresetIndex = recentPresets.findIndex(
    (preset) => (0, import_isSelectedPreset.isSelectedPreset)(segments, preset.segments)
  );
  if (recentPresetIndex > 0) {
    const oldPreset = old.recentPresets[recentPresetIndex];
    recentPresets.splice(recentPresetIndex, 1);
    recentPresets.unshift(oldPreset);
  }
  if (recentPresetIndex === -1) {
    const tempId = crypto.randomUUID();
    recentPresets.unshift({ id: tempId, segments });
    recentPresets.splice(10);
  }
  return {
    ...old,
    recentPresets
  };
}
const useSegments = () => {
  const store = (0, import_react.useContext)(import_SegmentsContext.SegmentsContext);
  const storeRef = (0, import_react.useRef)(store);
  const segments = (0, import_selector.useStoreSelector)(
    import_SegmentsContext.SegmentsContext,
    (state) => state
  );
  const segmentsRef = (0, import_react.useRef)(segments);
  const { getSegment, getAllSegments } = (0, import_SegmentsCache.useSegmentsCache)();
  const { dispatch } = (0, import_react.useContext)(import_SegmentsUIDispatchContext.UISegmentsDispatchContext);
  const getAllSegmentsRef = (0, import_react.useRef)(getAllSegments);
  const getSegmentRef = (0, import_react.useRef)(getSegment);
  const { mutate: addToRecentPresetsMutationFn } = (0, import_react_query.useMutation)(
    {
      mutationKey: ["add-recent-segment-preset"],
      mutationFn: (segments2) => (0, import_segments_presets.addRecentSegmentsPreset)(segments2),
      onMutate: async () => (0, import_queryHelper.runOptimisticMutation)(
        "getSegmentsPresets",
        (prevState) => updateCache(segments, prevState)
      ),
      onError: (error, _, context) => (0, import_queryHelper.handleFailedOptimisticMutation)("getSegmentsPresets", error, context),
      onSettled: () => (0, import_queryHelper.handleSettledOptimisticMutation)(
        ["getSegmentsPresets"],
        ["add-recent-segment-preset"]
      )
    },
    import_SegmentsCache.segmentsClient.queryClient
  );
  (0, import_react.useEffect)(() => {
    segmentsRef.current = segments;
  }, [segments]);
  (0, import_react.useEffect)(() => {
    storeRef.current = store;
  }, [store]);
  (0, import_react.useEffect)(() => {
    getAllSegmentsRef.current = getAllSegments;
  }, [getAllSegments]);
  (0, import_react.useEffect)(() => {
    getSegmentRef.current = getSegment;
  }, [getSegment]);
  const addSegment = (0, import_react.useCallback)(
    async (segment) => {
      const isAlreadyInStore = segmentsRef.current.find(
        (storeSegment) => storeSegment.id === segment.id
      );
      if (isAlreadyInStore) {
        return false;
      }
      try {
        const safeSegment = await import_SegmentsCache.segmentsClient.queryClient.fetchQuery({
          queryKey: ["getSegment", segment.id],
          queryFn: () => getSegmentRef.current(segment.id)
        });
        if (safeSegment) {
          storeRef.current.setState((state) => {
            const newState = [...state];
            newState.push({ id: segment.id, variables: segment.variables });
            addToRecentPresetsMutationFn(newState);
            return newState;
          });
        }
      } catch (e) {
        if ((0, import_is_node_environment.isDev)()) {
          console.warn(
            "SegmentSelector: The segment is not in the segments manager.",
            e
          );
        }
        dispatch({ type: "remove", payload: { id: segment.id } });
        return false;
      }
    },
    [dispatch, addToRecentPresetsMutationFn, getSegmentRef]
  );
  const removeSegment = (0, import_react.useCallback)((id) => {
    if (!id) {
      return;
    }
    storeRef.current.setState(
      (state) => state.filter((storeSegments) => id !== storeSegments.id)
    );
  }, []);
  const removeAllSegments = (0, import_react.useCallback)(
    () => storeRef.current.setState([]),
    []
  );
  const setSegments = (0, import_react.useCallback)(
    async (segments2, force) => {
      if (force) {
        storeRef.current.setState(segments2);
        addToRecentPresetsMutationFn(segments2);
        return;
      }
      if (import_SegmentsCache.segmentsClient.queryClient.getQueryData([
        "getAllSegments"
      ]) === void 0) {
        try {
          await import_SegmentsCache.segmentsClient.queryClient.fetchQuery({
            queryKey: ["getAllSegments"],
            queryFn: getAllSegmentsRef.current
          });
        } catch (error) {
          if ((0, import_is_node_environment.isDev)()) {
            console.warn(
              "SegmentSelector: Couldn't set segments due to a failed fetch.",
              error
            );
          }
          if (import_SegmentsCache.segmentsClient.hasAccessRights) {
            console.error(
              "SegmentSelector: Couldn't set segments due to a failed fetch.",
              { cause: error }
            );
            return;
          }
        }
      }
      if (!import_SegmentsCache.segmentsClient.hasAccessRights) {
        storeRef.current.setState(segments2);
        return;
      }
      const isValid = await (0, import_segmentValidationHelper.areUiSegmentsValid)(segments2);
      if (!isValid) {
        console.error(
          "SegmentSelector: Couldn't set segments due to an invalid selection."
        );
        return;
      }
      storeRef.current.setState(segments2);
      addToRecentPresetsMutationFn(segments2);
    },
    [addToRecentPresetsMutationFn]
  );
  return {
    segments,
    addSegment,
    removeSegment,
    removeAllSegments,
    setSegments
  };
};
