// @dynatrace/strato-components/filters/segment-selector/hooks/useTriggerProps.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useTriggerProps_exports = {};
__export(useTriggerProps_exports, {
  useTriggerProps: () => useTriggerProps
});
module.exports = __toCommonJS(useTriggerProps_exports);
var import_react_query = require("@tanstack/react-query");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_useSegments = require("./useSegments.js");
var import_useAriaLabelingProps = require("../../../core/hooks/useAriaLabelingProps.js");
var import_SegmentsCache = require("../contexts/SegmentsCache/SegmentsCache.js");
var import_segmentMessages = require("../segmentMessages.js");
var import_utils = require("../utils.js");
function useTriggerProps(variant, remainingProps) {
  const intl = (0, import_react_intl.useIntl)();
  const { segments } = (0, import_useSegments.useSegments)();
  const { getSegment } = (0, import_SegmentsCache.useSegmentsCache)();
  const safeId = segments[0] ? segments[0].id : void 0;
  const shouldFetchText = variant !== "compact" && segments.length === 1 && safeId !== void 0;
  const { data, isLoading } = (0, import_react_query.useQuery)(
    {
      queryKey: ["getSegment", safeId, shouldFetchText],
      queryFn: () => getSegment(safeId),
      enabled: shouldFetchText
    },
    import_SegmentsCache.segmentsClient.queryClient
  );
  const text = (0, import_react.useMemo)(() => {
    if (segments.length === 0 || !safeId) {
      return null;
    }
    if (variant === "compact") {
      return `${segments.length}`;
    }
    if (segments.length > 1) {
      return intl.formatMessage(
        {
          defaultMessage: "{length} segments",
          id: "7vkTbyAZNeY/qa9p",
          description: "Text for SegmentSelector Trigger when multiple segments are selected."
        },
        {
          length: segments.length
        }
      );
    }
    if (!data) {
      return `${safeId}`;
    }
    const amountOfVariables = segments[0]?.variables?.[0]?.values?.length ?? 0;
    if (amountOfVariables === 1) {
      return `${data.segment.name}: ${(0, import_utils.getReadableValue)(
        segments[0].variables?.[0].values[0],
        intl
      )}`;
    }
    if (amountOfVariables > 1) {
      return `${data.segment.name} (${amountOfVariables})`;
    }
    return `${data.segment.name}`;
  }, [data, intl, safeId, segments, variant]);
  const ariaLabelingProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(remainingProps);
  const ariaLabel = (0, import_react.useMemo)(() => {
    const ariaLabelProp = ariaLabelingProps["aria-label"];
    if (ariaLabelProp === void 0 && ariaLabelingProps["aria-labelledby"] === void 0) {
      return text ? void 0 : intl.formatMessage(import_segmentMessages.SEGMENTSELECTOR_MESSAGES.defaultAriaLabel);
    }
    const combinedAriaLabel = text ? ariaLabelProp?.concat(", ", text) : ariaLabelProp;
    return combinedAriaLabel;
  }, [ariaLabelingProps, intl, text]);
  const ariaProps = {
    ...ariaLabelingProps,
    "aria-label": ariaLabel,
    "aria-disabled": remainingProps["aria-disabled"]
  };
  return { isLoading: shouldFetchText ? isLoading : false, text, ariaProps };
}
