// @dynatrace/strato-design-tokens/index.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var index_exports = {};
__export(index_exports, {
  Animations: () => import_animations.default,
  AnimationsJs: () => import_animations_js.default,
  Borders: () => import_borders.default,
  BoxShadows: () => import_box_shadows.default,
  Breakpoints: () => import_breakpoints.default,
  BreakpointsJS: () => import_breakpoints_js.default,
  Colors: () => import_colors.default,
  Easings: () => import_easings.default,
  EasingsJs: () => import_easings_js.default,
  Elevations: () => import_elevations.default,
  Spacings: () => import_spacings.default,
  Timings: () => import_timings.default,
  TimingsJs: () => import_timings_js.default,
  Typography: () => import_typography.default,
  Variables: () => import_variables.default,
  VariablesDark: () => import_variables_dark.default
});
module.exports = __toCommonJS(index_exports);
var import_animations = __toESM(require("./animations/index.js"));
var import_animations_js = __toESM(require("./animations-js/index.js"));
var import_borders = __toESM(require("./borders/index.js"));
var import_box_shadows = __toESM(require("./box-shadows/index.js"));
var import_breakpoints = __toESM(require("./breakpoints/index.js"));
var import_breakpoints_js = __toESM(require("./breakpoints-js/index.js"));
var import_colors = __toESM(require("./colors/index.js"));
var import_easings = __toESM(require("./easings/index.js"));
var import_easings_js = __toESM(require("./easings-js/index.js"));
var import_elevations = __toESM(require("./elevations/index.js"));
var import_spacings = __toESM(require("./spacings/index.js"));
var import_timings = __toESM(require("./timings/index.js"));
var import_timings_js = __toESM(require("./timings-js/index.js"));
var import_typography = __toESM(require("./typography/index.js"));
var import_variables = __toESM(require("./variables/index.js"));
var import_variables_dark = __toESM(require("./variables-dark/index.js"));
