// @dynatrace/strato-components/charts/pie/components/InnerWrapper/components/InnerContainer/InnerContainer.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var InnerContainer_exports = {};
__export(InnerContainer_exports, {
  InnerContainer: () => InnerContainer
});
module.exports = __toCommonJS(InnerContainer_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_InnerContainer_css = require("./InnerContainer.sty.js");
var import_Flex = require("../../../../../../layouts/flex/Flex.js");
const InnerContainer = ({
  donutInner,
  height,
  width,
  children
}) => {
  const containerStyle = {
    top: `${height / 2 - donutInner / 2}px`,
    left: `${width / 2 - donutInner / 2}px`
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Flex.Flex,
    {
      justifyContent: "center",
      alignItems: "center",
      style: containerStyle,
      className: import_InnerContainer_css.container,
      children
    }
  );
};
