// @dynatrace/strato-geo/map/components/ConnectionLayer/ConnectionLayer.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ConnectionLayer_exports = {};
__export(ConnectionLayer_exports, {
  ConnectionLayer: () => ConnectionLayer
});
module.exports = __toCommonJS(ConnectionLayer_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_core = require("@dynatrace/strato-components/core");
var import_ConnectionLayerDirection = require("./ConnectionLayerDirection.js");
var import_ConnectionLayerLine = require("./ConnectionLayerLine.js");
var import_ConnectionLayerTooltip = require("./ConnectionLayerTooltip.js");
var import_parse_connection_data_to_geo_json = require("./utils/parse-connection-data-to-geo-json.js");
var import_constants = require("../../constants.js");
var import_use_layer_coloring_strategy = require("../../hooks/use-layer-coloring-strategy.js");
var import_use_legend_interaction_highlighting = require("../../hooks/use-legend-interaction-highlighting.js");
var import_use_legend_interaction_visibility = require("../../hooks/use-legend-interaction-visibility.js");
var import_use_resolve_color = require("../../hooks/use-resolve-color.js");
var import_use_tooltip_template = require("../../hooks/use-tooltip-template.js");
var import_ConnectionLayerTooltip2 = require("../../slots/ConnectionLayerTooltip.js");
var import_tooltip_type_guards = require("../../utils/tooltip-type-guards.js");
var import_TooltipWrapper = require("../TooltipWrapper.js");
const ConnectionLayer = (props) => {
  const [behavioralTrackingProps] = (0, import_core._useBehavioralTrackingProps)(props);
  const {
    data,
    layerId,
    children,
    curve,
    directionIndicator,
    line,
    thickness = import_constants.DEFAULT_LINE_THICKNESS
  } = props;
  const colorParser = (0, import_use_layer_coloring_strategy.useLayerColoringStrategy)();
  const connectionColor = (0, import_use_resolve_color.useResolveConnectionColor)(
    import_constants.DEFAULT_LINE_COLOR,
    props,
    colorParser
  );
  const sourceId = `source-${layerId}`;
  (0, import_use_legend_interaction_highlighting.useLegendInteractionHighlighting)(props, sourceId);
  (0, import_use_legend_interaction_visibility.useLegendInteractionVisibility)(props, layerId);
  const geoJsonData = (0, import_parse_connection_data_to_geo_json.parseConnectionDataToGeoJSON)(
    data,
    thickness,
    connectionColor,
    curve
  );
  const {
    handler: tooltipTemplate,
    hidden: tooltipHidden,
    symbolAlignment
  } = (0, import_use_tooltip_template.useTooltipTemplate)(children, import_ConnectionLayerTooltip2.ConnectionLayerTooltip, {
    extractProps: true
  });
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_ConnectionLayerLine.ConnectionLayerLine,
      {
        geoJsonData,
        layerId,
        sourceId,
        line,
        ...behavioralTrackingProps,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TooltipWrapper.TooltipWrapper, { layerId, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_ConnectionLayerTooltip.ConnectionLayerTooltip,
          {
            tooltipTemplate: (0, import_tooltip_type_guards.isConnectionTooltipTemplate)(tooltipTemplate) ? tooltipTemplate : void 0,
            hidden: tooltipHidden,
            symbolAlignment
          }
        ) })
      }
    ),
    directionIndicator && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_ConnectionLayerDirection.ConnectionLayerDirection,
      {
        geoJsonData,
        layerId,
        directionIndicator
      }
    )
  ] });
};
ConnectionLayer["displayName"] = "ConnectionLayer";
