// @dynatrace/strato-geo/map/components/ConnectionLayer/ConnectionLayerLine.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ConnectionLayerLine_exports = {};
__export(ConnectionLayerLine_exports, {
  ConnectionLayerLine: () => ConnectionLayerLine
});
module.exports = __toCommonJS(ConnectionLayerLine_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_maplibre = require("@vis.gl/react-maplibre");
var import_core = require("@dynatrace/strato-components/core");
var import_constants = require("../../constants.js");
const ConnectionLayerLine = (props) => {
  const { geoJsonData, layerId, sourceId, line, children, ...restProps } = props;
  const [behavioralTrackingProps] = (0, import_core._useBehavioralTrackingProps)(restProps);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_react_maplibre.Source, { id: sourceId, type: "geojson", data: geoJsonData, promoteId: "id", children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_react_maplibre.Layer,
      {
        type: "line",
        id: layerId,
        layout: {
          "line-cap": import_constants.DEFAULT_LINE_CAP,
          "line-join": import_constants.DEFAULT_LINE_JOIN
        },
        paint: {
          "line-color": [
            "case",
            ["boolean", ["feature-state", "hover"], false],
            ["get", "__lineHoveredColor"],
            ["get", "__lineColor"]
          ],
          "line-opacity": [
            "case",
            ["boolean", ["feature-state", "legendHover"], true],
            import_constants.SHAPE_OPACITY_DEFAULT,
            import_constants.SHAPE_OPACITY_DIMMED
          ],
          "line-width": ["get", "__lineWidth"],
          ...line && line === "dashed" ? { "line-dasharray": import_constants.DEFAULT_LINE_DASH_ARRAY } : null
        },
        ...behavioralTrackingProps
      }
    ),
    children
  ] });
};
ConnectionLayerLine["displayName"] = "ConnectionLayerLine";
