// @dynatrace/strato-components/charts/honeycomb/HoneycombChart.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var HoneycombChart_exports = {};
__export(HoneycombChart_exports, {
  HoneycombChart: () => HoneycombChart,
  _HoneycombChart: () => _HoneycombChart
});
module.exports = __toCommonJS(HoneycombChart_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_HoneycombChartLayout = require("./components/HoneycombChartLayout.js");
var import_constants = require("./constants.js");
var import_honeycomb_color_scheme_context = require("./context/honeycomb-color-scheme.context.js");
var import_honeycomb_formatter_context = require("./context/honeycomb-formatter.context.js");
var import_honeycomb_loading_context = require("./context/honeycomb-loading.context.js");
var import_honeycomb_shape_context = require("./context/honeycomb-shape.context.js");
var import_honeycomb_show_labels_context = require("./context/honeycomb-show-labels.context.js");
var import_honeycomb_truncation_mode_context = require("./context/honeycomb-truncation-mode.context.js");
var import_use_honeycomb_chart_config = require("./hooks/use-honeycomb-chart-config.js");
var import_honeycomb_base_scale_provider = require("./providers/honeycomb-base-scale.provider.js");
var import_honeycomb_configuration_provider = require("./providers/honeycomb-configuration.provider.js");
var import_honeycomb_dataset_type_provider = require("./providers/honeycomb-dataset-type.provider.js");
var import_honeycomb_filtered_data_provider = require("./providers/honeycomb-filtered-data.provider.js");
var import_honeycomb_metadata_provider = require("./providers/honeycomb-metadata.provider.js");
var import_honeycomb_store_provider = require("./providers/honeycomb-store.provider.js");
var import_DownloadCSV = require("./slots/DownloadCSV.js");
var import_HoneycombLegend = require("./slots/HoneycombLegend.js");
var import_Toolbar = require("./slots/Toolbar.js");
var import_Tooltip = require("./slots/Tooltip.js");
var import_build_and_download_csv = require("./utils/build-and-download-csv.js");
var import_normalize_tiles_data = require("./utils/normalize-tiles-data.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_Intent = require("../../core/slots/Intent/Intent.js");
var import_defaultChartValues_css = require("../../core/styles/defaultChartValues.sty.js");
var import_ChartSeriesActionContext = require("../core/components/series-action/ChartSeriesActionContext.js");
var import_get_toolbar_config_without_intents = require("../core/components/toolbar/utils/get-toolbar-config-without-intents.js");
var import_EmptyStateSlot = require("../core/slots/chart-states/EmptyStateSlot.js");
var import_ErrorStateSlot = require("../core/slots/chart-states/ErrorStateSlot.js");
var import_ColorRule = require("../core/slots/ColorRule.js");
var import_chart_types = require("../core/types/chart-types.js");
var import_coerceSizeValue = require("../core/utils/coerceSizeValue.js");
const COMPONENT_NAME = "HoneycombChart";
const normalizeHoneycombConfig = (config) => {
  const { toolbar: configuredToolbar, ...remainingConfig } = config;
  const toolbar = configuredToolbar ? (0, import_get_toolbar_config_without_intents.getToolbarConfigWithoutIntents)(configuredToolbar) : void 0;
  return { ...remainingConfig, toolbar };
};
const _HoneycombChart = (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    data,
    seriesActions,
    width,
    height,
    style: costumerStyle,
    className: customerClassNames,
    loading,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "data-testid": dataTestId
  } = props;
  const [behavioralTrackingProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(props);
  const containerRef = (0, import_react.useRef)(null);
  const chartWidth = (0, import_coerceSizeValue.coerceSizeValue)(width);
  const chartHeight = (0, import_coerceSizeValue.coerceSizeValue)(height) || import_constants.DEFAULT_HONEYCOMB_HEIGHT;
  const rootRef = (0, import_useMergeRefs.useMergeRefs)([
    forwardedRef,
    containerRef
  ]);
  const sanitizedData = (0, import_normalize_tiles_data.normalizeTilesData)(data);
  const config = (0, import_use_honeycomb_chart_config.useHoneycombChartConfig)(props);
  const formatterContext = {
    formatter: config.formatter,
    unit: config.unit
  };
  (0, import_react.useImperativeHandle)(forwardedRef, () => {
    const consumerConfig = normalizeHoneycombConfig(config);
    return {
      element: containerRef.current,
      downloadData: () => {
        (0, import_build_and_download_csv.buildAndDownloadCsv)(sanitizedData);
      },
      getConfig: () => {
        return JSON.stringify({
          type: import_chart_types.ChartType.Honeycomb,
          ...consumerConfig
        });
      }
    };
  });
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      ...behavioralTrackingProps,
      ref: rootRef,
      style: { ...costumerStyle, width: chartWidth, height: chartHeight },
      className: (0, import_clsx.default)(import_defaultChartValues_css.defaultChartSize, customerClassNames),
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "data-dt-component": COMPONENT_NAME,
      "aria-label": import_constants.HONEYCOMB_CHART_ARIA_LABEL,
      role: "img",
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_honeycomb_loading_context.HoneycombLoadingContext.Provider, { value: loading, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_honeycomb_configuration_provider.HoneycombConfigurationProvider, { config, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_honeycomb_shape_context.HoneycombShapeContext.Provider, { value: config.shape, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_honeycomb_color_scheme_context.HoneycombColorSchemeContext.Provider, { value: config.colorScheme, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_honeycomb_truncation_mode_context.HoneycombTruncationMode.Provider, { value: config.truncationMode, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_honeycomb_formatter_context.HoneycombFormatterContext.Provider, { value: formatterContext, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_honeycomb_show_labels_context.HoneycombShowLabelsContext.Provider,
        {
          value: config.showLabels,
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_honeycomb_filtered_data_provider.HoneycombFilteredDataProvider,
            {
              data: sanitizedData,
              config,
              children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_honeycomb_dataset_type_provider.HoneycombDatasetTypeProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_honeycomb_base_scale_provider.HoneycombBaseScaleProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_ChartSeriesActionContext.ChartSeriesActionContext.Provider,
                {
                  value: {
                    // We need a cast here as the provider expects unknown and the
                    // react context will not be able to carry the generic anyways.
                    chartSeriesActions: seriesActions
                  },
                  children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_honeycomb_store_provider.HoneycombStoreProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_honeycomb_metadata_provider.HoneycombMetadataProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_HoneycombChartLayout.HoneycombChartLayout, { ...props }) }) })
                }
              ) }) })
            }
          )
        }
      ) }) }) }) }) }) })
    }
  );
});
_HoneycombChart.displayName = import_constants.DISPLAY_NAME;
const HoneycombChart = Object.assign(_HoneycombChart, {
  EmptyState: import_EmptyStateSlot.EmptyStateSlot,
  ErrorState: import_ErrorStateSlot.ErrorStateSlot,
  Legend: import_HoneycombLegend.HoneycombLegend,
  Intent: import_Intent.Intent,
  ColorRule: import_ColorRule.ColorRule,
  Toolbar: import_Toolbar.Toolbar,
  DownloadCSV: import_DownloadCSV.HoneycombChartDownloadCSVSlot,
  Tooltip: import_Tooltip.Tooltip
});
