// @dynatrace/strato-components/charts/index.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var charts_exports = {};
__export(charts_exports, {
  AnnotationsChart: () => import_AnnotationsChart.AnnotationsChart,
  AnnotationsChartConfig: () => import_AnnotationsChartConfig_context.AnnotationsChartConfig,
  CSS_DOWNLOAD_FILENAME: () => import_constants.CSS_DOWNLOAD_FILENAME,
  CategoricalBarChart: () => import_CategoricalBarChartRoot.CategoricalBarChart,
  CategoricalBarChartConfig: () => import_useCategoricalBarChartConfig.CategoricalBarChartConfig,
  ChartInteractions: () => import_ChartInteractions.ChartInteractions,
  ChartSeriesAction: () => import_ChartSeriesAction.ChartSeriesAction,
  ChartToolbar: () => import_ChartToolbar.ChartToolbar,
  ChartTooltip: () => import_OverlayTooltip2.ChartTooltip,
  DEFAULT_TRUNCATION_MODE: () => import_constants.DEFAULT_TRUNCATION_MODE,
  DonutChart: () => import_DonutChart.DonutChart,
  DonutChartConfig: () => import_ChartConfigContext.DonutChartConfig,
  GaugeChart: () => import_GaugeChart.GaugeChart,
  GaugeChartConfig: () => import_GaugeChartConfigContext.GaugeChartConfig,
  HORIZONTAL_LABEL_LAYOUT_TYPE: () => import_types.HORIZONTAL_LABEL_LAYOUT_TYPE,
  HistogramAnnotations: () => import_HistogramChart.HistogramAnnotations,
  HistogramChart: () => import_HistogramChart.HistogramChart,
  HistogramChartConfig: () => import_HistogramChartConfig_context.HistogramChartConfig,
  HoneycombChart: () => import_HoneycombChart.HoneycombChart,
  HoneycombChartConfig: () => import_honeycomb_chart_config_context.HoneycombChartConfig,
  MeterBarChart: () => import_MeterBarChart.MeterBarChart,
  MeterBarChartConfig: () => import_MeterBarChartConfigContext.MeterBarChartConfig,
  MultiMeterBarChart: () => import_MultiMeterBarChart.MultiMeterBarChart,
  MultiMeterBarChartConfig: () => import_MultiMeterBarChartConfigContext.MultiMeterBarChartConfig,
  PieChart: () => import_PieChart.PieChart,
  PieChartConfig: () => import_ChartConfigContext.PieChartConfig,
  SharedChartInteractions: () => import_SharedChartInteractions.SharedChartInteractions,
  SingleValue: () => import_SingleValue.SingleValue,
  SingleValueConfig: () => import_use_single_value_config.SingleValueConfig,
  SingleValueGrid: () => import_SingleValueGrid.SingleValueGrid,
  Sparkline: () => import_Sparkline.Sparkline,
  TimeseriesAnnotations: () => import_TimeseriesChart.TimeseriesAnnotations,
  TimeseriesChart: () => import_TimeseriesChart.TimeseriesChart,
  TimeseriesChartConfig: () => import_TimeseriesChartConfig_context.TimeseriesChartConfig,
  TreeMap: () => import_TreeMap.TreeMap,
  TreeMapConfig: () => import_TreeMapConfigContext.TreeMapConfig,
  VERTICAL_LABEL_LAYOUT_TYPE: () => import_types.VERTICAL_LABEL_LAYOUT_TYPE,
  XYChart: () => import_XYChartRoot.XYChart,
  XYChartAnnotations: () => import_XYChartRoot.XYChartAnnotations,
  XYChartConfig: () => import_XYChartConfig_context.XYChartConfig,
  _AnnotationsAtom: () => import_AnnotationsAtom.AnnotationsAtom,
  _ChartGrid: () => import_ChartGridRoot.ChartGrid,
  _ChartLayout: () => import_ChartLayout.ChartLayout,
  _ChartLegend: () => import_ChartLegend.ChartLegend,
  _ChartToolbarRenderer: () => import_ChartToolbarRenderer.ChartToolbarRenderer,
  _ChartTooltip: () => import_ChartTooltip.ChartTooltip,
  _ChartTooltipAtoms: () => import_ChartTooltip_config2.ChartTooltipAtoms,
  _ChartTooltipSlots: () => import_ChartTooltip_config.ChartTooltipSlots,
  _ChartTooltipTemplate: () => import_ChartTooltip_config3.ChartTooltipTemplate,
  _DurationChip: () => import_DurationChip.DurationChip,
  _EmptyState: () => import_EmptyState.EmptyState,
  _ErrorState: () => import_ErrorState.ErrorState,
  _EventItem: () => import_EventItem.EventItem,
  _FPS: () => import_constants.SIXTY_FPS,
  _GEO_DEFAULTS: () => import_defaults.GEO_DEFAULTS,
  _GeoItem: () => import_GeoItem.GeoItem,
  _InteractionsOverlay: () => import_InteractionsOverlay.InteractionsOverlay,
  _KeyValue: () => import_KeyValue.KeyValue,
  _LegendSizeContext: () => import_LegendSizeContext.LegendSizeContext,
  _MarkerAtom: () => import_MarkerAtom.MarkerAtom,
  _MeterBarChartThresholdLegend: () => import_MeterBarChartThresholdLegend.MeterBarChartThresholdLegend,
  _OverlayProvider: () => import_OverlayContext.OverlayProvider,
  _OverlayTooltip: () => import_OverlayTooltip.OverlayTooltip,
  _OverlayTooltipCustomStory: () => import_OverlayTooltipWIPStory.OverlayTooltipCustomStory,
  _OverlayTooltipHeaderStory: () => import_OverlayTooltipWIPStory.OverlayTooltipHeaderStory,
  _OverlayTooltipProvider: () => import_TooltipContext.TooltipProvider,
  _OverlayTooltipTruncationStory: () => import_OverlayTooltipWIPStory.OverlayTooltipTruncationStory,
  _OverlayTooltipWIPStory: () => import_OverlayTooltipWIPStory.OverlayTooltipWIPStory,
  _RangeLegend: () => import_RangeLegend.RangeLegend,
  _SeriesItem: () => import_SeriesItem.SeriesItem,
  _TextHeader: () => import_TextHeader.TextHeader,
  _ThresholdBar: () => import_ThresholdBar.ThresholdBar,
  _ThresholdIndicator: () => import_ThresholdIndicator.ThresholdIndicator,
  _TimeFrameHeader: () => import_TimeFrameHeader.TimeFrameHeader,
  _TrackAtom: () => import_TrackAtom.TrackAtom,
  _XYChartTooltipBody: () => import_XYChartTooltipBody.XYChartTooltipBody,
  _XYChartTooltipHeader: () => import_XYChartTooltipHeader.XYChartTooltipHeader,
  _XYChartTooltipTemplate: () => import_XYChartTooltipTemplate.XYChartTooltipTemplate,
  _automaticColoringMap: () => import_color_palettes_definitions2.automaticColoringMap,
  _automaticColoringOptions: () => import_color_palette2.automaticColoringOptions,
  _buildCategoricalItemsFromScale: () => import_build_categorical_items_from_scale.buildCategoricalItemsFromScale,
  _buildLegendItemsFromRange: () => import_build_legend_items_from_range.buildLegendItemsFromRange,
  _buildQuantizeScaleFromColorPalette: () => import_build_quantize_scale_from_color_palette.buildQuantizeScaleFromColorPalette,
  _buildQuantizedRangeItems: () => import_build_quantized_range_items.buildQuantizedRangeItems,
  _buildThresholdScaleFromRanges: () => import_build_threshold_scale_from_ranges.buildThresholdScaleFromRanges,
  _coerceSizeValue: () => import_coerceSizeValue.coerceSizeValue,
  _colorPaletteMap: () => import_color_palettes_definitions.colorPaletteMap,
  _divergingColoringOptions: () => import_color_palette2.divergingColoringOptions,
  _getCanvasColor: () => import_canvas_color.getCanvasColor,
  _getLegendOrientation: () => import_orientation.getLegendOrientation,
  _getMetadataMetricsUnitForFieldName: () => import_get_metadata_metrics_unit_for_field_name._getMetadataMetricsUnitForFieldName,
  _getTabIndexKey: () => import_a11yToolbarControls2.getTabIndexKey,
  _getTabIndexValue: () => import_a11yToolbarControls.getTabIndexValue,
  _getTickFormatterMethod: () => import_get_tick_formatter_method.getTickFormatterMethod,
  _isCustomColorPalette: () => import_color_palette.isCustomColorPalette,
  _sequentialColoringOptions: () => import_color_palette2.sequentialColoringOptions,
  _themedColoringOptions: () => import_color_palette2.themedColoringOptions,
  _truncateText: () => import_text_truncation.truncateText,
  _useA11yToolbarControls: () => import_useA11yToolbarControls.useA11yToolbarControls,
  _useA11yToolbarTabIndex: () => import_useA11yToolbarTabIndex.useA11yToolbarTabIndex,
  _useAutoLegendRefresh: () => import_useAutoLegendRefresh.useAutoLegendRefresh,
  _useChartActionsMenu: () => import_useChartActionsMenu.useChartActionsMenu,
  _useChartLayoutLoading: () => import_useChartLayoutLoading.useChartLayoutLoading,
  _useChartLayoutOrientation: () => import_useChartLayoutOrientation.useChartLayoutOrientation,
  _useChartLayoutSize: () => import_useChartLayoutSize.useChartLayoutSize,
  _useDefaultConfig: () => import_useDefaultConfig.useDefaultConfig,
  _useDefaultLabels: () => import_useDefaultLabels.useDefaultLabels,
  _useGraphSize: () => import_useGraphSize.useGraphSize,
  _useGridPlotSize: () => import_useGridPlotSize.useGridPlotSize,
  _useLegendSize: () => import_LegendSizeContext2.useLegendSize,
  _useMousePosition: () => import_useMousePosition.useMousePosition,
  _useObserveContainerSize: () => import_useObserveContainerSize.useObserveContainerSize,
  _useOverlayChart: () => import_OverlayContext.useOverlay,
  _useOverlayTooltipActions: () => import_useTooltipActions.useTooltipActions,
  _useOverlayTooltipDefaultLabels: () => import_useDefaultLabels2.useDefaultLabels,
  _useOverlayTooltipReducer: () => import_useTooltipReducer.useTooltipReducer,
  _useOverlayTooltipStore: () => import_TooltipContext.useTooltip,
  _useThrottledResizeObserver: () => import_useThrottledResizeObserver.useThrottledResizeObserver,
  checkTimeseriesFormat: () => import_check_timeseries_format.checkTimeseriesFormat,
  colorPaletteModeOptions: () => import_color_palette2.colorPaletteModeOptions,
  colorPaletteOptions: () => import_color_palette2.colorPaletteOptions,
  convertQueryResultToTimeseries: () => import_convert_to_timeseries.convertQueryResultToTimeseries,
  convertToFieldEntries: () => import_getFieldsArrayForRecordIndex.convertToFieldEntries,
  convertToTimeseries: () => import_convert_to_timeseries.convertToTimeseries,
  convertToTimeseriesBand: () => import_convert_to_timeseries_band.convertToTimeseriesBand,
  getDimensions: () => import_get_dimensions.getDimensions,
  getFieldsArrayForRecordIndex: () => import_getFieldsArrayForRecordIndex.getFieldsArrayForRecordIndex,
  getTypesForRecordIndex: () => import_getFieldsArrayForRecordIndex.getTypesForRecordIndex,
  multiMeterBarChartTooltipVariantOptions: () => import_multi_meter_bar_chart.multiMeterBarChartTooltipVariantOptions,
  pointsDisplayOptions: () => import_canvas_shape.pointsDisplayOptions
});
module.exports = __toCommonJS(charts_exports);
var import_InteractionsOverlay = require("./core/components/interactions-overlay/InteractionsOverlay.js");
var import_AnnotationsChartConfig_context = require("./annotations-chart/contexts/AnnotationsChartConfig.context.js");
var import_AnnotationsChart = require("./annotations-chart/AnnotationsChart.js");
var import_CategoricalBarChartRoot = require("./categorical-bar/CategoricalBarChartRoot.js");
var import_useCategoricalBarChartConfig = require("./categorical-bar/hooks/useCategoricalBarChartConfig.js");
var import_AnnotationsAtom = require("./core/components/annotations/AnnotationsAtom.js");
var import_MarkerAtom = require("./core/components/annotations/slots/MarkerAtom.js");
var import_TrackAtom = require("./core/components/annotations/slots/TrackAtom.js");
var import_types = require("./core/components/axes/types.js");
var import_ChartGridRoot = require("./core/components/chart-grid/ChartGridRoot.js");
var import_ChartLayout = require("./core/components/chart-layout/ChartLayout.js");
var import_orientation = require("./core/components/chart-layout/utils/orientation.js");
var import_EmptyState = require("./core/components/chart-states/EmptyState.js");
var import_ErrorState = require("./core/components/chart-states/ErrorState.js");
var import_ChartLegend = require("./core/components/legend/categorical/ChartLegend.js");
var import_RangeLegend = require("./core/components/legend/range/RangeLegend.js");
var import_LegendSizeContext = require("./core/components/legend/shared/context/LegendSizeContext.js");
var import_ChartSeriesAction = require("./core/components/series-action/ChartSeriesAction.js");
var import_useChartActionsMenu = require("./core/components/series-action/hooks/useChartActionsMenu.js");
var import_ThresholdBar = require("./core/components/threshold-bar/ThresholdBar.js");
var import_ThresholdIndicator = require("./core/components/threshold-bar/indicators/ThresholdIndicator.js");
var import_ChartToolbarRenderer = require("./core/components/toolbar/ChartToolbarRenderer.js");
var import_ChartTooltip = require("./core/components/tooltip/ChartTooltip/ChartTooltip.js");
var import_ChartTooltip_config = require("./core/components/tooltip/types/ChartTooltip.config.js");
var import_ChartTooltip_config2 = require("./core/components/tooltip/types/ChartTooltip.config.js");
var import_ChartTooltip_config3 = require("./core/components/tooltip/types/ChartTooltip.config.js");
var import_GeoItem = require("./core/components/tooltip/ChartTooltip/components/GeoItem.js");
var import_SeriesItem = require("./core/components/tooltip/ChartTooltip/components/SeriesItem/SeriesItem.js");
var import_TimeFrameHeader = require("./core/components/tooltip/ChartTooltip/components/TimeFrameHeader/TimeFrameHeader.js");
var import_EventItem = require("./core/components/tooltip/ChartTooltip/components/EventItem/EventItem.js");
var import_KeyValue = require("./core/components/tooltip/ChartTooltip/components/KeyValue.js");
var import_TextHeader = require("./core/components/tooltip/ChartTooltip/components/TextHeader.js");
var import_DurationChip = require("./core/components/tooltip/ChartTooltip/components/DurationChip.js");
var import_SharedChartInteractions = require("./core/components/shared-chart-interactions/SharedChartInteractions.js");
var import_OverlayTooltip = require("./core/components/overlay-tooltip/OverlayTooltip.js");
var import_OverlayTooltip2 = require("./core/components/overlay-tooltip/OverlayTooltip.js");
var import_defaults = require("./core/components/overlay-tooltip/defaults.js");
var import_useDefaultConfig = require("./core/components/overlay-tooltip/hooks/useDefaultConfig.js");
var import_useDefaultLabels = require("./core/components/overlay-tooltip/hooks/useDefaultLabels.js");
var import_OverlayContext = require("./core/components/overlay-tooltip/contexts/OverlayContext.js");
var import_TooltipContext = require("./core/components/overlay-tooltip/contexts/TooltipContext.js");
var import_useTooltipReducer = require("./core/components/overlay-tooltip/useTooltipReducer.js");
var import_useTooltipActions = require("./core/components/overlay-tooltip/useTooltipActions.js");
var import_useDefaultLabels2 = require("./core/components/overlay-tooltip/hooks/useDefaultLabels.js");
var import_OverlayTooltipWIPStory = require("./core/components/overlay-tooltip/_to-be-removed/_OverlayTooltipWIPStory.js");
var import_constants = require("./core/constants.js");
var import_a11yToolbarControls = require("./core/utils/a11yToolbarControls.js");
var import_a11yToolbarControls2 = require("./core/utils/a11yToolbarControls.js");
var import_build_quantized_range_items = require("./core/utils/build-quantized-range-items.js");
var import_build_categorical_items_from_scale = require("./core/utils/build-categorical-items-from-scale.js");
var import_build_legend_items_from_range = require("./core/utils/build-legend-items-from-range.js");
var import_build_quantize_scale_from_color_palette = require("./core/utils/build-quantize-scale-from-color-palette.js");
var import_build_threshold_scale_from_ranges = require("./core/utils/build-threshold-scale-from-ranges.js");
var import_coerceSizeValue = require("./core/utils/coerceSizeValue.js");
var import_color_palette = require("./core/types/color-palette.js");
var import_useA11yToolbarControls = require("./core/hooks/useA11yToolbarControls.js");
var import_useA11yToolbarTabIndex = require("./core/hooks/useA11yToolbarTabIndex.js");
var import_useThrottledResizeObserver = require("./core/hooks/useThrottledResizeObserver.js");
var import_useObserveContainerSize = require("./core/hooks/useObserveContainerSize.js");
var import_canvas_color = require("./core/components/canvas/utils/canvas-color.js");
var import_canvas_shape = require("./core/components/canvas/types/canvas-shape.js");
var import_useGraphSize = require("./core/components/chart-layout/hooks/useGraphSize.js");
var import_useChartLayoutOrientation = require("./core/components/chart-layout/hooks/useChartLayoutOrientation.js");
var import_useChartLayoutLoading = require("./core/components/chart-layout/hooks/useChartLayoutLoading.js");
var import_useAutoLegendRefresh = require("./core/components/chart-layout/hooks/useAutoLegendRefresh.js");
var import_useGridPlotSize = require("./core/components/chart-grid/hooks/useGridPlotSize.js");
var import_useChartLayoutSize = require("./core/components/chart-layout/hooks/useChartLayoutSize.js");
var import_LegendSizeContext2 = require("./core/components/legend/shared/context/LegendSizeContext.js");
var import_get_tick_formatter_method = require("./core/utils/get-tick-formatter-method.js");
var import_color_palette2 = require("./core/types/color-palette.js");
var import_ChartToolbar = require("./core/slots/toolbar/ChartToolbar.js");
var import_HistogramChart = require("./histogram/HistogramChart.js");
var import_HistogramChartConfig_context = require("./histogram/contexts/HistogramChartConfig.context.js");
var import_HoneycombChart = require("./honeycomb/HoneycombChart.js");
var import_honeycomb_chart_config_context = require("./honeycomb/context/honeycomb-chart-config.context.js");
var import_MultiMeterBarChart = require("./meter-bar/multi-meter-bar/MultiMeterBarChart.js");
var import_MultiMeterBarChartConfigContext = require("./meter-bar/multi-meter-bar/context/MultiMeterBarChartConfigContext.js");
var import_multi_meter_bar_chart = require("./meter-bar/multi-meter-bar/types/multi-meter-bar-chart.js");
var import_MeterBarChart = require("./meter-bar/single-meter-bar/MeterBarChart.js");
var import_MeterBarChartThresholdLegend = require("./meter-bar/single-meter-bar/components/MeterBarChartThresholdLegend/MeterBarChartThresholdLegend.js");
var import_MeterBarChartConfigContext = require("./meter-bar/single-meter-bar/context/MeterBarChartConfigContext.js");
var import_PieChart = require("./pie/PieChart.js");
var import_DonutChart = require("./pie/DonutChart.js");
var import_ChartConfigContext = require("./pie/context/ChartConfigContext.js");
var import_SingleValue = require("./single-value/SingleValue.js");
var import_SingleValueGrid = require("./single-value-grid/SingleValueGrid.js");
var import_use_single_value_config = require("./single-value/hooks/use-single-value-config.js");
var import_Sparkline = require("./sparkline/Sparkline.js");
var import_TimeseriesChart = require("./timeseries/TimeseriesChart.js");
var import_TimeseriesChartConfig_context = require("./timeseries/context/TimeseriesChartConfig.context.js");
var import_ChartInteractions = require("./core/slots/chart-interactions/ChartInteractions.js");
var import_convert_to_timeseries = require("./timeseries/utils/convert-to-timeseries/convert-to-timeseries.js");
var import_convert_to_timeseries_band = require("./timeseries/utils/convert-to-timeseries/convert-to-timeseries-band.js");
var import_getFieldsArrayForRecordIndex = require("./timeseries/utils/convert-to-timeseries/utils/getFieldsArrayForRecordIndex.js");
var import_check_timeseries_format = require("./timeseries/utils/convert-to-timeseries/utils/check-timeseries-format.js");
var import_get_dimensions = require("./timeseries/utils/convert-to-timeseries/utils/get-dimensions.js");
var import_get_metadata_metrics_unit_for_field_name = require("./timeseries/utils/convert-to-timeseries/utils/get-metadata-metrics-unit-for-field-name.js");
var import_XYChartRoot = require("./xy-chart/XYChartRoot.js");
var import_XYChartConfig_context = require("./xy-chart/context/XYChartConfig.context.js");
var import_XYChartTooltipTemplate = require("./xy-chart/components/tooltip/XYChartTooltipTemplate/XYChartTooltipTemplate.js");
var import_XYChartTooltipHeader = require("./xy-chart/components/tooltip/XYChartTooltipHeader/XYChartTooltipHeader.js");
var import_XYChartTooltipBody = require("./xy-chart/components/tooltip/XYChartTooltipBody/XYChartTooltipBody.js");
var import_TreeMap = require("./tree-map/TreeMap.js");
var import_TreeMapConfigContext = require("./tree-map/context/TreeMapConfigContext.js");
var import_useMousePosition = require("./core/hooks/useMousePosition.js");
var import_text_truncation = require("./core/utils/text-truncation.js");
var import_GaugeChart = require("./gauge/GaugeChart.js");
var import_GaugeChartConfigContext = require("./gauge/contexts/GaugeChartConfigContext.js");
var import_color_palettes_definitions = require("./core/types/color-palettes-definitions.js");
var import_color_palettes_definitions2 = require("./core/types/color-palettes-definitions.js");
