// @dynatrace/strato-components/charts/timeseries/components/displays/band/accessors.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var accessors_exports = {};
__export(accessors_exports, {
  definedAccessor: () => definedAccessor,
  xAxisAccessor: () => xAxisAccessor,
  y0: () => y0,
  y0AxisAccessor: () => y0AxisAccessor,
  y1: () => y1,
  y1AxisAccessor: () => y1AxisAccessor
});
module.exports = __toCommonJS(accessors_exports);
var import_safe_get = require("../../../../core/utils/safe-get.js");
var import_timeseries = require("../../../utils/timeseries.js");
const definedAccessor = (dataPoint, datapointsMap) => !(0, import_safe_get.safeGet)(datapointsMap, dataPoint).isGap;
const xAxisAccessor = (xScale, datapointsMap) => (dataPoint) => {
  const { start, end, center } = (0, import_safe_get.safeGet)(datapointsMap, dataPoint);
  return xScale((0, import_timeseries.getDatapointTime)(start, end, center)) ?? 0;
};
const y0AxisAccessor = (scale, dataPointsMap) => (dataPoint) => {
  const dataPointMetadata = (0, import_safe_get.safeGet)(dataPointsMap, dataPoint);
  const value = dataPointMetadata.y0 ?? 0;
  return scale(value);
};
const y1AxisAccessor = (scale, dataPointsMap) => (dataPoint) => {
  const dataPointMetadata = (0, import_safe_get.safeGet)(dataPointsMap, dataPoint);
  const value = dataPointMetadata.y1 ?? 0;
  return scale(value);
};
function y1(yScaleByUnit, seriesMetadata, datapointsMap) {
  return (dataPoint) => y1AxisAccessor(
    yScaleByUnit(seriesMetadata.unit),
    datapointsMap
  )(dataPoint) ?? 0;
}
function y0(yScaleByUnit, seriesMetadata, datapointsMap) {
  return (dataPoint) => y0AxisAccessor(
    yScaleByUnit(seriesMetadata.unit),
    datapointsMap
  )(dataPoint) ?? 0;
}
