// @dynatrace/strato-components/charts/timeseries/components/displays/band/BandChartDisplay.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var BandChartDisplay_exports = {};
__export(BandChartDisplay_exports, {
  BandChartDisplay: () => BandChartDisplay
});
module.exports = __toCommonJS(BandChartDisplay_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_accessors = require("./accessors.js");
var import_band_style = require("./band-style.js");
var import_AreaCanvas = require("../../../../core/components/canvas/components/AreaCanvas.js");
var import_useGridPlotSize = require("../../../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_constants = require("../../../../core/constants.js");
var import_safe_get = require("../../../../core/utils/safe-get.js");
var import_useSeriesCurve = require("../../../hooks/useSeriesCurve.js");
var import_useTimeseriesFilteredData = require("../../../hooks/useTimeseriesFilteredData.js");
var import_useYScales = require("../../../hooks/useYScales.js");
var import_yAxis = require("../../../utils/yAxis.js");
const BandChartDisplay = ({
  series: timeseries,
  xScale
}) => {
  const yScale = (0, import_useYScales.useYScales)();
  const { width, height } = (0, import_useGridPlotSize.useGridPlotSize)();
  const { seriesMap, datapointsMap } = (0, import_useTimeseriesFilteredData.useTimeseriesFilteredData)();
  const curve = (0, import_useSeriesCurve.useSeriesCurve)();
  const bandSeries = (0, import_react.useMemo)(() => [...timeseries].reverse(), [timeseries]);
  const yScaleByUnit = (unit) => (0, import_yAxis.getYScaleByUnit)(unit, yScale);
  const displayScatterPlot = (_datapoints, _index) => false;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, { children: bandSeries.map((series) => {
    const seriesMetadata = (0, import_safe_get.safeGet)(seriesMap, series);
    const { color, stroke, opacity } = (0, import_band_style.getBandStyle)(
      seriesMetadata,
      series.strokeOnly
    );
    const strokeColor = stroke ?? "";
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_AreaCanvas.AreaCanvas,
      {
        data: series.datapoints,
        canvasWidth: width,
        canvasHeight: height,
        x: (0, import_accessors.xAxisAccessor)(xScale, datapointsMap),
        y1: (0, import_accessors.y1)(yScaleByUnit, seriesMetadata, datapointsMap),
        y0: (0, import_accessors.y0)(yScaleByUnit, seriesMetadata, datapointsMap),
        color: () => color,
        strokeColor: () => strokeColor,
        defined: (dataPoint) => (0, import_accessors.definedAccessor)(dataPoint, datapointsMap),
        opacity,
        displayScatterPlot,
        strokeWidth: import_constants.DEFAULT_STROKE_WIDTH_BAND_CHART,
        curve,
        enableDotOverflow: false
      },
      seriesMetadata.id
    );
  }) });
};
BandChartDisplay["displayName"] = "BandChartDisplay";
