// @dynatrace/strato-components/charts/core/components/toolbar/components/ChartDataMenu.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ChartDataMenu_exports = {};
__export(ChartDataMenu_exports, {
  ChartDataMenu: () => ChartDataMenu
});
module.exports = __toCommonJS(ChartDataMenu_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react2 = require("react");
var import_react_intl = require("react-intl");
var import_spacings = __toESM(require("@dynatrace/strato-design-tokens/spacings"));
var import_strato_icons = require("@dynatrace/strato-icons");
var import_AsyncIntents = require("./AsyncIntents.js");
var import_ChartDataMenu_css = require("./ChartDataMenu.sty.js");
var import_ToolbarMenu_css = require("./ToolbarMenu.sty.js");
var import_Button = require("../../../../../buttons/button/Button.js");
var import_Menu = require("../../../../../navigation/menu/Menu.js");
var import_useA11yToolbarTabIndex = require("../../../hooks/useA11yToolbarTabIndex.js");
var import_useEffectOnValueChange = require("../../../hooks/useEffectOnValueChange.js");
var import_messages = require("../messages.js");
var import_divider = require("./divider.js");
const ChartDataMenu = ({
  onMenuEnter,
  onMenuLeave,
  onMenuOpen,
  menuAlignment = "end",
  isDownloadEnabled,
  handleDownloadDataToolbar,
  intents = [],
  showDivider,
  buttonMode = "standard",
  isDragging,
  intentsAsyncTemplate
}) => {
  const intentButtonRef = (0, import_react2.useRef)(null);
  const menuContentRef = (0, import_react2.useRef)(null);
  const tabIndex = (0, import_useA11yToolbarTabIndex.useA11yToolbarTabIndex)("intent");
  const intl = (0, import_react_intl.useIntl)();
  const [isOpen, setIsOpen] = (0, import_react2.useState)(false);
  const isInlineMode = buttonMode === "inline";
  const handleOnMouseEnter = () => {
    onMenuEnter?.();
  };
  (0, import_useEffectOnValueChange.useEffectOnValueChange)(() => {
    onMenuOpen(isOpen);
  }, [isOpen]);
  (0, import_react2.useEffect)(() => {
    if (tabIndex === 0 && intentButtonRef.current) {
      intentButtonRef.current?.focus();
    }
  }, [intentButtonRef, tabIndex]);
  (0, import_react2.useEffect)(() => {
    if (isDragging !== void 0 && isDragging && isOpen) {
      setIsOpen(false);
    }
  }, [isDragging, isOpen]);
  const handleOnMouseLeave = () => {
    onMenuLeave?.();
  };
  const handleToolbarMenuClick = () => {
    setIsOpen((prevState) => !prevState);
  };
  const handleOnClickOutsideOverlay = (0, import_react2.useCallback)(
    (e) => {
      const menuContainsTarget = menuContentRef.current?.contains(
        e.target
      );
      if (isOpen && !menuContainsTarget) {
        setIsOpen(false);
      }
    },
    [setIsOpen, isOpen, menuContentRef]
  );
  const handleOnKeyUp = (0, import_react2.useCallback)(
    (e) => {
      if (e.key === "Escape") {
        setIsOpen(false);
      }
    },
    [setIsOpen]
  );
  const handleOnKeyDown = (e) => {
    const nativeEvent = e.nativeEvent;
    if (nativeEvent.code === "Enter") {
      setIsOpen((prevState) => !prevState);
    }
  };
  (0, import_react2.useEffect)(() => {
    window.addEventListener("click", handleOnClickOutsideOverlay);
    window.addEventListener("keyup", handleOnKeyUp);
    return () => {
      window.removeEventListener("click", handleOnClickOutsideOverlay);
      window.removeEventListener("keyup", handleOnKeyUp);
    };
  }, [handleOnClickOutsideOverlay, handleOnKeyUp]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    "div",
    {
      className: (0, import_ToolbarMenu_css.styledToolbarMenuContainer)({
        inline: isInlineMode
      }),
      onMouseEnter: handleOnMouseEnter,
      onMouseLeave: handleOnMouseLeave,
      onClick: handleToolbarMenuClick,
      onKeyDown: handleOnKeyDown,
      ref: menuContentRef,
      children: [
        showDivider && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_divider.VerticalDivider, {}),
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Menu.Menu, { open: isOpen, children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Trigger, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_Button.Button,
            {
              width: import_spacings.default.Size28,
              className: isInlineMode ? import_ChartDataMenu_css.triggerButton : void 0,
              variant: "default",
              size: "condensed",
              ref: intentButtonRef,
              tabIndex,
              children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.DotMenuIcon, {})
            }
          ) }),
          /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Menu.Menu.Content, { alignment: menuAlignment, children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Group, { className: import_ToolbarMenu_css.toolbarMenuGroup, children: isDownloadEnabled && handleDownloadDataToolbar ? /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Menu.Menu.Item, { onSelect: handleDownloadDataToolbar, children: [
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.DownloadIcon, {}) }),
              intl.formatMessage(import_messages.TOOLBAR_MESSAGES.downloadDataButton)
            ] }) : null }),
            intents.length ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Group, { className: import_ToolbarMenu_css.toolbarMenuGroup, children: intents.map((item, key) => /* @__PURE__ */ (0, import_react.createElement)(import_Menu.Menu.Intent, { ...item, key: item.children?.toString() || key }, item.icon && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Prefix, { children: item.icon }), item.children)) }) : null,
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_AsyncIntents.AsyncIntents,
              {
                intentsAsyncTemplate,
                shouldRun: isOpen
              }
            )
          ] })
        ] })
      ]
    }
  );
};
