// @dynatrace/strato-components/charts/core/components/toolbar/components/Control.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Control_exports = {};
__export(Control_exports, {
  Control: () => Control
});
module.exports = __toCommonJS(Control_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_Control_css = require("./Control.sty.js");
var import_Button = require("../../../../../buttons/button/Button.js");
var import_KeyboardShortcutTooltip = require("../../../../../content/keyboard-shortcut-tooltip/KeyboardShortcutTooltip.js");
var import_useMergeRefs = require("../../../../../core/hooks/useMergeRefs.js");
var import_Divider = require("../../../../../layouts/divider/Divider.js");
var import_Menu = require("../../../../../navigation/menu/Menu.js");
var import_Tooltip = require("../../../../../overlays/tooltip/Tooltip.js");
var import_ChartToolbarRenderer_css = require("../ChartToolbarRenderer.sty.js");
var import_useToolbarContext = require("../hooks/useToolbarContext.js");
const Control = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    prefixIcon,
    tooltipText,
    testId,
    selected = false,
    onClick,
    onKeyDown,
    tabIndex,
    disabled = false,
    hideTooltip = false,
    tooltipKeyShortcut,
    placement = "auto",
    showWhenCollapsed = false,
    divider = "none",
    className: consumerClassName,
    style: consumerStyle,
    "data-dt-name": dtName,
    "data-dt-properties": dtProperties
  } = props;
  const { isMenuElement, isExpanded, didDrag, resetDidDrag } = (0, import_useToolbarContext.useToolbarContext)();
  const localRef = (0, import_react.useRef)(null);
  const rootRef = (0, import_useMergeRefs.useMergeRefs)([
    forwardedRef,
    localRef
  ]);
  (0, import_react.useEffect)(() => {
    if (tabIndex === 0) {
      localRef.current?.focus();
    }
  }, [tabIndex]);
  const handleClick = (event) => {
    event.stopPropagation();
    if (didDrag) {
      resetDidDrag?.();
      return;
    }
    if (!disabled) {
      onClick(event);
    }
  };
  const handleSelect = (event) => {
    event.stopPropagation();
    if (!disabled) {
      onClick(event);
    }
  };
  const handleKeyDown = (event) => {
    if (!disabled) {
      onKeyDown?.(event);
    }
  };
  const styledButton = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Button.Button,
    {
      ref: rootRef,
      className: (0, import_clsx.default)((0, import_Control_css.buttonStyle)({ disabled, selected }), consumerClassName),
      style: consumerStyle,
      variant: "default",
      tabIndex,
      "data-testid": testId,
      "aria-label": tooltipText,
      onKeyDown: handleKeyDown,
      onMouseUp: handleClick,
      size: "condensed",
      "data-dt-children-name": dtName,
      "data-dt-properties": dtProperties,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: prefixIcon })
    }
  );
  if (isMenuElement) {
    if (placement === "toolbar" || placement === "auto" && isExpanded) {
      return null;
    }
    const horizontalDivider = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Divider.Divider,
      {
        className: import_ChartToolbarRenderer_css.styledHorizontalDivider,
        orientation: "horizontal",
        as: "hr"
      }
    );
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
      divider === "left" && horizontalDivider,
      /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
        import_Menu.Menu.Item,
        {
          className: (0, import_Control_css.menuItemStyles)({ disabled, selected }),
          onSelect: handleSelect,
          onKeyDown: handleKeyDown,
          "data-dt-children-name": dtName,
          "data-dt-properties": dtProperties,
          children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Prefix, { children: prefixIcon }),
            /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { className: import_Control_css.styledMenuItemContent, children: [
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { className: import_Control_css.styledText, children: tooltipText }),
              /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("span", { className: (0, import_Control_css.styledShortcut)({ disabled }), children: [
                "\xA0",
                tooltipKeyShortcut
              ] })
            ] })
          ]
        }
      ),
      divider === "right" && horizontalDivider
    ] });
  }
  if (placement === "menu" || !isExpanded && !showWhenCollapsed) {
    return null;
  }
  const verticalDivider = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Divider.Divider, { className: import_ChartToolbarRenderer_css.styledVerticalDivider, orientation: "vertical", as: "hr" });
  let computedControlButton = styledButton;
  if (tooltipText) {
    if (tooltipKeyShortcut) {
      computedControlButton = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_KeyboardShortcutTooltip.KeyboardShortcutTooltip,
        {
          keys: tooltipKeyShortcut,
          text: tooltipText,
          disabled: hideTooltip,
          shortcutDisabled: disabled,
          children: styledButton
        }
      );
    } else {
      computedControlButton = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Tooltip.Tooltip, { text: tooltipText, disabled: hideTooltip, children: styledButton });
    }
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    (divider === "left" || placement === "toolbar" && !isExpanded) && verticalDivider,
    computedControlButton,
    divider === "right" && verticalDivider
  ] });
});
Control.displayName = "ChartToolbar.Control";
