// @dynatrace/strato-components/charts/core/components/toolbar/components/divider.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var divider_exports = {};
__export(divider_exports, {
  HorizontalDivider: () => HorizontalDivider,
  VerticalDivider: () => VerticalDivider
});
module.exports = __toCommonJS(divider_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_Divider = require("../../../../../layouts/divider/Divider.js");
var import_ChartToolbarRenderer_css = require("../ChartToolbarRenderer.sty.js");
const HorizontalDivider = () => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
  import_Divider.Divider,
  {
    className: import_ChartToolbarRenderer_css.styledHorizontalDivider,
    orientation: "horizontal",
    as: "hr",
    style: { width: "inherit" }
  }
);
const VerticalDivider = () => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Divider.Divider, { className: import_ChartToolbarRenderer_css.styledVerticalDivider, orientation: "vertical", as: "hr" });
