// @dynatrace/strato-components/charts/core/components/toolbar/components/ToolbarMenu.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ToolbarMenu_exports = {};
__export(ToolbarMenu_exports, {
  ToolbarMenu: () => ToolbarMenu
});
module.exports = __toCommonJS(ToolbarMenu_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_Button = require("../../../../../buttons/button/Button.js");
var import_Menu = require("../../../../../navigation/menu/Menu.js");
var import_Tooltip = require("../../../../../overlays/tooltip/Tooltip.js");
var import_useA11yToolbarTabIndex = require("../../../hooks/useA11yToolbarTabIndex.js");
var import_useToolbarContext = require("../hooks/useToolbarContext.js");
var import_messages = require("../messages.js");
var import_ToolbarMenu_css = require("./ToolbarMenu.sty.js");
function renderMenuItems(children) {
  return import_react.Children.map(children, (child) => {
    if ((0, import_react.isValidElement)(child)) {
      if (child.type === import_react.Fragment) {
        return renderMenuItems(child.props.children);
      }
      return (0, import_react.cloneElement)(child, {
        ...child.props
      });
    }
    return child;
  });
}
const ToolbarMenu = (props) => {
  const {
    children,
    onMenuOpen,
    isOpen,
    activeModeIcon,
    triggerLabel,
    onMenuEnter,
    onMenuLeave,
    menuAlignment = "start",
    onCloseAutoFocus,
    "data-dt-name": dtName,
    "data-dt-properties": dtProperties
  } = props;
  const [isHoverChartToolsIcon, setIsHoverChartToolsIcon] = (0, import_react.useState)(false);
  const menuTriggerButtonRef = (0, import_react.useRef)(null);
  const tabIndex = (0, import_useA11yToolbarTabIndex.useA11yToolbarTabIndex)("explore");
  const intl = (0, import_react_intl.useIntl)();
  const { isExpanded } = (0, import_useToolbarContext.useToolbarContext)();
  const menuItems = (0, import_react.useMemo)(() => renderMenuItems(children), [children]);
  (0, import_react.useEffect)(() => {
    if (tabIndex === 0 && menuTriggerButtonRef.current) {
      menuTriggerButtonRef.current?.focus();
    }
  }, [menuTriggerButtonRef, tabIndex]);
  const chartActionsButton = (0, import_react.useMemo)(() => {
    const tooltipText = isExpanded ? triggerLabel ?? intl.formatMessage(import_messages.TOOLBAR_MESSAGES.changeMode) : intl.formatMessage(import_messages.TOOLBAR_MESSAGES.chartTools);
    const handleMenuOpenChange = (isOpen2) => setIsHoverChartToolsIcon(isOpen2);
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Tooltip.Tooltip,
      {
        text: tooltipText,
        open: isHoverChartToolsIcon,
        onOpenChange: handleMenuOpenChange,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Trigger, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
          import_Button.Button,
          {
            variant: "default",
            size: "condensed",
            ref: menuTriggerButtonRef,
            tabIndex,
            "data-dt-name": dtName,
            "data-dt-properties": dtProperties,
            children: [
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: activeModeIcon ?? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.CrosshairIcon, {}) }),
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Suffix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ChevronDownSmallIcon, {}) })
            ]
          }
        ) })
      }
    );
  }, [
    isHoverChartToolsIcon,
    isExpanded,
    activeModeIcon,
    triggerLabel,
    intl,
    tabIndex,
    dtName,
    dtProperties
  ]);
  const handleOnMouseEnter = () => {
    onMenuEnter?.();
  };
  const handleOnMouseLeave = () => {
    onMenuLeave?.();
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      className: (0, import_ToolbarMenu_css.styledToolbarMenuContainer)({
        inline: props.toolbarDisplayMode === "inline"
      }),
      onMouseEnter: handleOnMouseEnter,
      onMouseLeave: handleOnMouseLeave,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Menu.Menu, { onOpenChange: onMenuOpen, open: isOpen, children: [
        chartActionsButton,
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_Menu.Menu.Content,
          {
            alignment: menuAlignment,
            onCloseAutoFocus,
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Group, { className: import_ToolbarMenu_css.toolbarMenuGroup, children: menuItems })
          }
        )
      ] })
    }
  );
};
