// @dynatrace/strato-components/layouts/app-header/ActionItems.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ActionItems_exports = {};
__export(ActionItems_exports, {
  ActionItems: () => ActionItems
});
module.exports = __toCommonJS(ActionItems_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react2 = require("react");
var import_react_intl = require("react-intl");
var import_react_is = require("react-is");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_ActionButton = require("./ActionButton.js");
var import_ActionItemGroup = require("./ActionItemGroup.js");
var import_Button = require("../../buttons/button/Button.js");
var import_useContainerSize = require("../../core/hooks/useContainerSize.js");
var import_is_slot = require("../../core/utils/is-slot.js");
var import_Menu = require("../../navigation/menu/Menu.js");
var import_Tooltip = require("../../overlays/tooltip/Tooltip.js");
var import_Divider = require("../divider/Divider.js");
var import_Flex = require("../flex/Flex.js");
function getSlots(children) {
  const actions = [];
  const actionGroups = {};
  const custom = [];
  let groupCounter = 0;
  for (const child of import_react2.Children.toArray(children)) {
    if (!(0, import_react_is.isElement)(child)) {
      continue;
    }
    const childWithKey = (0, import_react2.cloneElement)(child, {
      key: child.key ?? crypto.randomUUID()
    });
    if ((0, import_is_slot.isSlot)(child, import_ActionItemGroup.ActionItemGroup)) {
      groupCounter++;
      actionGroups[groupCounter] = [];
      const actionGroup = actionGroups[groupCounter];
      const validChildren = import_react2.Children.toArray(child.props.children).filter(
        (validChild) => (0, import_is_slot.isSlot)(validChild, import_Tooltip.Tooltip) && (0, import_is_slot.isSlot)(validChild.props.children, import_ActionButton.ActionButton) || (0, import_is_slot.isSlot)(validChild, import_ActionButton.ActionButton)
      );
      validChildren.forEach((validChild) => actionGroup.push(validChild));
    } else if ((0, import_ActionButton.isAppHeaderActionButton)(child)) {
      actions.push(childWithKey);
    } else if ((0, import_is_slot.isSlot)(child, import_Tooltip.Tooltip) && (0, import_is_slot.isSlot)(child.props.children, import_ActionButton.ActionButton)) {
      actions.push(child);
    } else {
      custom.push(childWithKey);
    }
  }
  return { actions, actionGroups, custom };
}
function useCollapse() {
  const containerRef = (0, import_react2.useRef)(null);
  const contentRef = (0, import_react2.useRef)(null);
  const containerSize = (0, import_useContainerSize.useContainerSize)(containerRef);
  const contentSize = (0, import_useContainerSize.useContainerSize)(contentRef);
  return {
    isOverflowing: contentSize.width > containerSize.width,
    containerRef,
    contentRef
  };
}
const ActionItems = (0, import_react2.forwardRef)(({ children, className, style, "data-testid": dataTestId, ...rest }, ref) => {
  const slots = (0, import_react2.useMemo)(() => getSlots(children), [children]);
  const collapse = useCollapse();
  const intl = (0, import_react_intl.useIntl)();
  const hasActions = slots.actions.length > 0;
  const hasGroups = Object.entries(slots.actionGroups).length > 0;
  const isCollapsedIntoMenu = collapse.isOverflowing && (hasActions || hasGroups);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Flex.Flex,
      {
        ref,
        alignItems: "center",
        justifyContent: "flex-end",
        flexGrow: 1,
        flexShrink: 0,
        flexBasis: hasActions || hasGroups ? 0 : "auto",
        minWidth: 0,
        style,
        className,
        "data-testid": dataTestId,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_Flex.Flex,
          {
            ...rest,
            ref: collapse.containerRef,
            justifyContent: "flex-end",
            alignItems: "center",
            flexGrow: 1,
            minWidth: 0,
            hidden: collapse.isOverflowing,
            style: {
              visibility: collapse.isOverflowing ? "hidden" : void 0,
              flexBasis: collapse.isOverflowing ? 0 : void 0
            },
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
              import_Flex.Flex,
              {
                gap: 4,
                style: { minWidth: "max-content" },
                ref: collapse.contentRef,
                alignItems: "center",
                children: [
                  collapse.isOverflowing ? null : slots.custom,
                  slots.actions.map((action) => {
                    if ((0, import_is_slot.isSlot)(action, import_ActionButton.ActionButton)) {
                      const { prefixIcon, children: children2, showLabel, ...rest2 } = action.props;
                      return /* @__PURE__ */ (0, import_react.createElement)(import_Button.Button, { ...rest2, key: action.key, variant: "emphasized" }, prefixIcon ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: prefixIcon }) : null, children2);
                    } else if ((0, import_is_slot.isSlot)(action, import_Tooltip.Tooltip) && (0, import_is_slot.isSlot)(action.props.children, import_ActionButton.ActionButton)) {
                      return action;
                    }
                    return null;
                  }),
                  hasActions && hasGroups ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                    import_Divider.Divider,
                    {
                      orientation: "vertical",
                      flexItem: false,
                      style: { height: 20 }
                    }
                  ) : null,
                  Object.entries(slots.actionGroups).map(
                    ([key, actions], actionGroupIndex) => /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_react2.Fragment, { children: [
                      actionGroupIndex !== 0 && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                        import_Divider.Divider,
                        {
                          orientation: "vertical",
                          flexItem: false,
                          style: { height: 20 }
                        }
                      ),
                      actions.map((action) => {
                        if ((0, import_is_slot.isSlot)(action, import_ActionButton.ActionButton)) {
                          const { prefixIcon, children: children2, showLabel, ...rest2 } = action.props;
                          return /* @__PURE__ */ (0, import_react.createElement)(import_Button.Button, { ...rest2, key: action.key, variant: "emphasized" }, prefixIcon ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: prefixIcon }) : null, children2);
                        } else if ((0, import_is_slot.isSlot)(action, import_Tooltip.Tooltip) && (0, import_is_slot.isSlot)(action.props.children, import_ActionButton.ActionButton)) {
                          return action;
                        }
                        return null;
                      })
                    ] }, key)
                  )
                ]
              }
            )
          }
        )
      }
    ),
    collapse.isOverflowing ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Flex.Flex, { gap: 4, alignItems: "center", children: slots.custom }) : null,
    isCollapsedIntoMenu ? /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Menu.Menu, { children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Trigger, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_Button.Button,
        {
          variant: "default",
          "aria-label": intl.formatMessage({
            id: "L/D6X7Ee5PYWzl3i",
            defaultMessage: "Show more actions"
          }),
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.DotMenuIcon, {}) })
        }
      ) }),
      /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Menu.Menu.Content, { children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Group, { children: slots.actions.map((action) => {
          if ((0, import_is_slot.isSlot)(action, import_ActionButton.ActionButton)) {
            const { prefixIcon, children: children2, showLabel, onClick, ...rest2 } = action.props;
            return /* @__PURE__ */ (0, import_react.createElement)(
              import_Menu.Menu.Item,
              {
                ...rest2,
                key: action.key,
                onSelect: onClick ? onClick : () => void 0
              },
              prefixIcon ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Prefix, { children: prefixIcon }) : null,
              children2
            );
          } else if ((0, import_is_slot.isSlot)(action, import_Tooltip.Tooltip) && (0, import_is_slot.isSlot)(action.props.children, import_ActionButton.ActionButton)) {
            return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
              import_Menu.Menu.Item,
              {
                onSelect: action.props.children.props.onClick ? action.props.children.props.onClick : () => void 0,
                ...action.props.children,
                children: [
                  action.props.children.props.prefixIcon ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Prefix, { children: action.props.children.props.prefixIcon }) : null,
                  action.props.children.props.children,
                  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.ItemTooltip, { children: action.props.text })
                ]
              }
            );
          }
          return null;
        }) }),
        Object.entries(slots.actionGroups).map(([groupId, actions]) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Group, { children: actions.map((action) => {
          if ((0, import_is_slot.isSlot)(action, import_ActionButton.ActionButton)) {
            const {
              prefixIcon,
              children: children2,
              showLabel,
              onClick,
              ...rest2
            } = action.props;
            return /* @__PURE__ */ (0, import_react.createElement)(
              import_Menu.Menu.Item,
              {
                ...rest2,
                key: action.key,
                onSelect: onClick ? onClick : () => void 0
              },
              prefixIcon ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Prefix, { children: prefixIcon }) : null,
              children2
            );
          } else if ((0, import_is_slot.isSlot)(action, import_Tooltip.Tooltip) && (0, import_is_slot.isSlot)(action.props.children, import_ActionButton.ActionButton)) {
            return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
              import_Menu.Menu.Item,
              {
                onSelect: action.props.children.props.onClick ? action.props.children.props.onClick : () => void 0,
                ...action.props.children,
                children: [
                  action.props.children.props.prefixIcon ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Prefix, { children: action.props.children.props.prefixIcon }) : null,
                  action.props.children.props.children,
                  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.ItemTooltip, { children: action.props.text })
                ]
              }
            );
          }
          return null;
        }) }, groupId))
      ] })
    ] }) : null
  ] });
});
ActionItems.displayName = "AppHeader.ActionItems";
