// @dynatrace/strato-components/layouts/app-header/AppHeader.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var AppHeader_exports = {};
__export(AppHeader_exports, {
  AppHeader: () => AppHeader
});
module.exports = __toCommonJS(AppHeader_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_ActionButton = require("./ActionButton.js");
var import_ActionItemGroup = require("./ActionItemGroup.js");
var import_ActionItems = require("./ActionItems.js");
var import_AppIcon = require("./app-icon/AppIcon.js");
var styles = __toESM(require("./AppHeader.sty.js"));
var import_Logo = require("./Logo.js");
var import_Menus = require("./Menus.js");
var import_Navigation = require("./Navigation.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_OverlayStackingProvider = require("../../core/providers/OverlayStackingProvider.js");
var import_is_node_environment = require("../../core/utils/is-node-environment.js");
var import_is_slot = require("../../core/utils/is-slot.js");
var import_Divider = require("../divider/Divider.js");
var import_HelpMenu = require("../help-menu/HelpMenu.js");
const COMPONENT_NAME = "AppHeader";
function getItemSlots(children) {
  const slots = {
    navigation: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(AppHeader.Navigation, {})
  };
  for (const child of import_react.Children.toArray(children)) {
    if ((0, import_is_slot.isSlot)(child, import_ActionItems.ActionItems)) {
      slots.actions = child;
      continue;
    }
    if ((0, import_is_slot.isSlot)(child, import_HelpMenu.HelpMenu)) {
      slots.menus = /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
        slots.menus,
        child
      ] });
      continue;
    }
    if ((0, import_is_slot.isSlot)(child, import_Menus.Menus)) {
      slots.menus = /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
        child,
        slots.menus
      ] });
      continue;
    }
    if ((0, import_is_slot.isSlot)(child, import_Navigation.Navigation)) {
      slots.navigation = child;
      continue;
    }
    if ((0, import_is_node_environment.isDev)()) {
      console.warn("The used child is not supported by the AppHeader.");
    }
  }
  return slots;
}
const AppHeaderRoot = (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const {
      children,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...remainingProps
    } = props;
    const slots = (0, import_react.useMemo)(() => getItemSlots(children), [children]);
    const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "header",
      {
        ...restProps,
        ...behavioralTrackingProps,
        ref: forwardedRef,
        className: (0, import_clsx.default)(styles.header, consumerClassName),
        style: consumerStyle,
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        "data-dt-component": COMPONENT_NAME,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_OverlayStackingProvider.OverlayStackingProvider, { children: [
          slots.navigation,
          slots.actions,
          slots.actions && slots.menus ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_Divider.Divider,
            {
              orientation: "vertical",
              flexItem: false,
              style: { height: 20 }
            }
          ) : null,
          slots.menus
        ] })
      }
    );
  }
);
AppHeaderRoot.displayName = "AppHeader";
const AppHeader = Object.assign(AppHeaderRoot, {
  // DEPRECATION: [APPDEV-14264] BREAKING-CHANGE: AppHeader: Remove old NavItems API.
  /**
   * @deprecated use {@link AppHeader.Logo}
   */
  AppNavLink: import_Logo.Logo,
  // DEPRECATION: [APPDEV-14264] BREAKING-CHANGE: AppHeader: Remove old NavItems API.
  /**
   * @deprecated use {@link AppHeader.Navigation}
   */
  NavItems: import_Navigation.Navigation,
  // DEPRECATION: [APPDEV-14264] BREAKING-CHANGE: AppHeader: Remove old NavItems API.
  /**
   * @deprecated use {@link AppHeader.NavigationItem}
   */
  NavItem: import_Navigation.NavigationItem,
  AppIcon: import_AppIcon.AppIcon,
  ActionItems: import_ActionItems.ActionItems,
  ActionItemGroup: import_ActionItemGroup.ActionItemGroup,
  ActionButton: import_ActionButton.ActionButton,
  Logo: import_Logo.Logo,
  Menus: import_Menus.Menus,
  Navigation: import_Navigation.Navigation,
  NavigationItem: import_Navigation.NavigationItem
});
