// @dynatrace/strato-components/layouts/app-header/Navigation.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Navigation_exports = {};
__export(Navigation_exports, {
  Navigation: () => Navigation,
  NavigationItem: () => NavigationItem
});
module.exports = __toCommonJS(Navigation_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_clsx = __toESM(require("clsx"));
var import_react2 = require("react");
var import_react_intl = require("react-intl");
var import_react_is = require("react-is");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_Logo = require("./Logo.js");
var styles = __toESM(require("./Navigation.sty.js"));
var import_Button = require("../../buttons/button/Button.js");
var import_useContainerSize = require("../../core/hooks/useContainerSize.js");
var import_is_slot = require("../../core/utils/is-slot.js");
var import_Menu = require("../../navigation/menu/Menu.js");
var import_Flex = require("../flex/Flex.js");
function getSlots(children) {
  const linksProps = [];
  let logo;
  for (const child of import_react2.Children.toArray(children)) {
    if (!(0, import_react_is.isElement)(child)) {
      continue;
    }
    if ((0, import_is_slot.isSlot)(child, NavigationItem)) {
      const key = child.key ?? crypto.randomUUID();
      const propsWithKey = Object.assign({}, child.props, { key });
      linksProps.push(propsWithKey);
    } else if ((0, import_is_slot.isSlot)(child, import_Logo.Logo)) {
      logo = child;
    }
  }
  return { linksProps, logo };
}
function useCollapsingNav() {
  const containerRef = (0, import_react2.useRef)(null);
  const contentRef = (0, import_react2.useRef)(null);
  const containerSize = (0, import_useContainerSize.useContainerSize)(containerRef);
  const contentSize = (0, import_useContainerSize.useContainerSize)(contentRef);
  return {
    isOverflowing: contentSize.width > containerSize.width,
    containerRef,
    contentRef
  };
}
const Navigation = (0, import_react2.forwardRef)(
  (props, ref) => {
    const {
      "data-testid": dataTestId,
      style,
      className,
      children,
      ...rest
    } = props;
    const slots = (0, import_react2.useMemo)(() => getSlots(children), [children]);
    const collapsingNav = useCollapsingNav();
    const intl = (0, import_react_intl.useIntl)();
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Menu.Menu, { children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Trigger, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_Button.Button,
          {
            "aria-label": intl.formatMessage({
              defaultMessage: "Open app navigation",
              id: "WCZAUtmjwNCm5iwa",
              description: "Label for the menu trigger to open app-header navigation."
            }),
            style: {
              display: collapsingNav.isOverflowing ? void 0 : "none"
            },
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.MenuIcon, {}) })
          }
        ) }),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Content, { children: slots.linksProps.map((linkProps) => {
          const {
            isSelected,
            disabled,
            className: navItemClassName,
            ...rest2
          } = linkProps;
          return /* @__PURE__ */ (0, import_react.createElement)(
            import_Menu.Menu.Link,
            {
              ...rest2,
              key: linkProps.key,
              className: (0, import_clsx.default)(
                styles.menuLink,
                {
                  [styles.isSelected]: isSelected,
                  [styles.isDisabled]: disabled
                },
                navItemClassName
              ),
              disabled
            }
          );
        }) })
      ] }),
      /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
        import_Flex.Flex,
        {
          ...rest,
          ref,
          gap: 4,
          alignItems: "center",
          flexShrink: 1,
          minWidth: 0,
          style,
          className,
          "data-testid": dataTestId,
          children: [
            slots.logo || /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Logo.Logo, {}),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_Flex.Flex,
              {
                ref: collapsingNav.containerRef,
                flexGrow: 1,
                minWidth: 0,
                hidden: collapsingNav.isOverflowing,
                style: {
                  visibility: collapsingNav.isOverflowing ? "hidden" : void 0,
                  flexBasis: collapsingNav.isOverflowing ? 0 : void 0
                },
                children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  import_Flex.Flex,
                  {
                    as: "nav",
                    gap: 4,
                    className: styles.inlineNav,
                    ref: collapsingNav.contentRef,
                    "aria-label": intl.formatMessage({
                      defaultMessage: "App navigation",
                      id: "oA42meteX3Ww5Zo5"
                    }),
                    children: slots.linksProps.map((linkProps) => {
                      const {
                        isSelected,
                        disabled,
                        as = "a",
                        className: navItemClassName,
                        ...rest2
                      } = linkProps;
                      return /* @__PURE__ */ (0, import_react.createElement)(
                        import_Button.Button,
                        {
                          ...rest2,
                          key: linkProps.key,
                          as,
                          className: (0, import_clsx.default)(
                            styles.button,
                            {
                              [styles.isSelected]: isSelected,
                              [styles.isDisabled]: disabled
                            },
                            navItemClassName
                          ),
                          disabled
                        }
                      );
                    })
                  }
                )
              }
            )
          ]
        }
      )
    ] });
  }
);
Navigation.displayName = "AppHeader.Navigation";
const NavigationItem = (0, import_react2.forwardRef)(
  (props, forwardedRef) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { ref: forwardedRef, ...props })
);
NavigationItem.displayName = "AppHeader.NavigationItem";
