// @dynatrace-sdk/client-app-settings/esm/index.js
/**
 * @license
 * Copyright 2023-2025 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// packages/client/app-settings/src/lib/apis/app-settings-objects-api.ts
import {
  httpClient as defaultHttpClient,
  isHttpClientAbortError,
  isHttpClientResponseError
} from "@dynatrace-sdk/http-client";
import {
  ApiClientError,
  ClientRequestError as ClientRequestError2,
  InvalidResponseError,
  isClientRequestError,
  isInvalidResponseError
} from "@dynatrace-sdk/shared-errors";

// packages/client/app-settings/src/lib/error-envelopes/api-gateway-errors-handler.ts
import { ApiGatewayError } from "@dynatrace-sdk/shared-errors";
async function apiGatewayErrorsHandler(response) {
  if (response.headers?.["dynatrace-response-source"] === "API Gateway") {
    const responseBody = await response.body("json");
    throw new ApiGatewayError(response, responseBody);
  }
}

// packages/client/app-settings/src/lib/error-envelopes/app-settings-error-envelope-error.ts
import { ClientRequestError } from "@dynatrace-sdk/shared-errors";
var AppSettingsErrorEnvelopeError = class extends ClientRequestError {
  isAppSettingsErrorEnvelopeError = true;
};
function isAppSettingsErrorEnvelopeError(e) {
  return e?.isApiClientError === true && e?.isClientRequestError === true && e?.isAppSettingsErrorEnvelopeError === true && e instanceof Error;
}

// packages/client/app-settings/src/lib/error-envelopes/get-error-message.ts
function serializeData(data) {
  try {
    return JSON.stringify(data);
  } catch (e) {
    return String(data);
  }
}
function getMessagesFromErrorDetails(details) {
  const messages = [];
  Object.entries(details).forEach(([name, data]) => {
    if (!data)
      return;
    const serializedData = serializeData(data);
    switch (name) {
      case "missingScopes":
        messages.push(`Missing scopes: ${serializedData}`);
        break;
      default:
        messages.push(`${name}: ${serializedData}`);
    }
  }, []);
  return messages;
}
function getErrorMessage(errorBody, defaultMessage) {
  const error = errorBody;
  const msg = error?.error?.message || defaultMessage;
  const details = error?.error?.details || {};
  return [msg, ...getMessagesFromErrorDetails(details)].join(". ");
}

// packages/client/app-settings/src/lib/utils/transformations.ts
var defaultRuleConfig = {
  force: false,
  datetime: false
};
function mergeKeys(key1, key2) {
  if (!key1)
    return key2;
  if (!key2)
    return key1;
  return `${key1}.${key2}`;
}
function matchKey(key, rules, cache) {
  if (cache.has(key)) {
    return cache.get(key);
  }
  for (const rule of rules) {
    const pattern = Array.isArray(rule) ? rule[0] : rule;
    const config = Array.isArray(rule) ? rule[1] : defaultRuleConfig;
    if (key === pattern) {
      const result = [pattern, config];
      cache.set(key, result);
      return result;
    }
    let i = 0;
    let j = 0;
    const keyLen = key.length;
    const patternLen = pattern.length;
    let isMatch = true;
    while (i < keyLen && j < patternLen) {
      let keyPartEnd = key.indexOf(".", i);
      let patternPartEnd = pattern.indexOf(".", j);
      if (keyPartEnd === -1)
        keyPartEnd = keyLen;
      if (patternPartEnd === -1)
        patternPartEnd = patternLen;
      const keyPart = key.slice(i, keyPartEnd);
      const patternPart = pattern.slice(j, patternPartEnd);
      if (patternPart !== "*" && patternPart !== "**" && keyPart !== patternPart) {
        isMatch = false;
        break;
      }
      i = keyPartEnd + 1;
      j = patternPartEnd + 1;
    }
    if (isMatch) {
      if (j < patternLen) {
        const remaining = pattern.slice(j);
        if (remaining !== "**" && remaining !== "*") {
          isMatch = false;
        }
      }
      if (i < keyLen && pattern.slice(j - 1) !== "**") {
        isMatch = false;
      }
    }
    if (isMatch) {
      const result = [pattern, config];
      cache.set(key, result);
      return result;
    }
  }
  cache.set(key, null);
  return null;
}
var potentialMatchForRoot = (pattern, key) => pattern.startsWith(key) || pattern.startsWith("*") || pattern.startsWith("**");
var potentialMatch = (pattern, key) => pattern.startsWith(key) || pattern.includes("*") || pattern.includes("**");
function partialMatch(key, keyPatterns) {
  const keyPatternsLen = keyPatterns.length;
  if (!key.includes(".")) {
    for (let i = 0; i < keyPatternsLen; i++) {
      const pattern = keyPatterns[i];
      if (potentialMatchForRoot(pattern, key))
        return true;
    }
    return false;
  }
  let dotIndex = -1;
  let currentKeyEnd = 0;
  while ((dotIndex = key.indexOf(".", currentKeyEnd)) !== -1) {
    const currentKey = key.slice(0, dotIndex);
    const isRoot = !currentKey.includes(".");
    for (let i = 0; i < keyPatternsLen; i++) {
      const pattern = keyPatterns[i];
      if (isRoot) {
        if (potentialMatchForRoot(pattern, currentKey))
          return true;
      } else if (potentialMatch(pattern, currentKey))
        return true;
    }
    currentKeyEnd = dotIndex + 1;
  }
  for (let i = 0; i < keyPatternsLen; i++) {
    const pattern = keyPatterns[i];
    if (potentialMatch(pattern, key))
      return true;
  }
  return false;
}
var dateTimePattern = /^(\d{4}-\d{2}-\d{2})(T\d{2}:\d{2}:\d{2}(?:\.\d+)?(Z|[+-]\d{2}:\d{2})?)?$/i;
function isDateLike(value) {
  if (typeof value === "string") {
    return dateTimePattern.test(value);
  }
  return false;
}
function transformValue(value, rules, direction) {
  const { force, datetime } = rules;
  if (direction === "to") {
    return force || isDateLike(value) ? new Date(value) : value;
  }
  if (value instanceof Date) {
    const iso = value.toISOString();
    return datetime ? iso : iso.split("T")[0];
  }
  return value;
}
function walk(obj, direction, rules, rulesCache, keyPatterns, currentKey = "") {
  if (rules.length <= 0) {
    return obj;
  }
  if (!partialMatch(currentKey, keyPatterns)) {
    return obj;
  }
  if (Array.isArray(obj)) {
    const match = matchKey(currentKey, rules, rulesCache);
    obj.forEach((item, idx) => {
      if (typeof item === "object" && item !== null) {
        walk(item, direction, rules, rulesCache, keyPatterns, currentKey);
      } else if (match) {
        obj[idx] = transformValue(obj[idx], match[1], direction);
      }
    });
  }
  if (typeof obj === "object" && obj !== null && !Array.isArray(obj)) {
    for (const key of Object.keys(obj)) {
      const mergedKey = mergeKeys(currentKey, key);
      const match = matchKey(mergedKey, rules, rulesCache);
      if (match) {
        obj[key] = transformValue(obj[key], match[1], direction);
      }
      walk(obj[key], direction, rules, rulesCache, keyPatterns, mergedKey);
    }
  }
  return obj;
}
function flattenKeyPatterns(keyPatterns) {
  return keyPatterns.map((k) => Array.isArray(k) ? k[0] : k);
}
function transform(direction, object, rules) {
  const flatKeyPatterns = flattenKeyPatterns(rules);
  const rulesCache = /* @__PURE__ */ new Map();
  return walk(object, direction, rules, rulesCache, flatKeyPatterns);
}
function transformRequest(object, rules) {
  return transform("from", object, rules);
}
function transformResponse(object, rules) {
  return transform("to", object, rules);
}

// packages/client/app-settings/src/lib/utils/url-helpers.ts
var encodeQueryParam = (key, value) => {
  const encodedKey = encodeURIComponent(key);
  return `${encodedKey}=${encodeURIComponent(typeof value === "number" ? value : String(value))}`;
};
var addExplodedArrayQueryParam = (query, key) => {
  const arrayValue = query[key];
  return arrayValue.map((value) => encodeQueryParam(key, value)).join("&");
};
var addNonExplodedArrayQueryParams = (query, key) => {
  const encodedKey = encodeURIComponent(key);
  const encodedParamsList = query[key].map((value) => encodeURIComponent(typeof value === "number" ? value : String(value))).join(",");
  return `${encodedKey}=${encodedParamsList}`;
};
var addQueryParam = (query, key) => encodeQueryParam(key, query[key]);
var arrayQueryParams = (query, key, explode) => {
  return explode ? addExplodedArrayQueryParam(query, key) : addNonExplodedArrayQueryParams(query, key);
};
var toQueryString = (rawQuery, flags = {}) => {
  const query = rawQuery || {};
  const keys = Object.keys(query).filter((key) => typeof query[key] !== "undefined");
  const queryString = keys.map(
    (key) => Array.isArray(query[key]) ? arrayQueryParams(query, key, flags.explode?.hasOwnProperty(key) ? flags.explode[key] : true) : addQueryParam(query, key)
  ).join("&");
  return queryString ? `?${queryString}` : "";
};

// packages/client/app-settings/src/lib/apis/app-settings-objects-api.ts
var AppSettingsObjectsClient = class {
  httpClient;
  shouldTransformDates = true;
  /**
   * @param {HttpClient} httpClientImplementation - You can provide custom http client as a parameter to constructor of a AppSettingsObjectsClient. Custom http client should implement {HttpClient} interface.
   * @param {ApiClientOptions} [options] - Optional configuration for SDK Client.
   * @example
   * const appSettingsObjectsClientNodeJs = new AppSettingsObjectsClient(yourCustomImplementation);
   */
  constructor(httpClientImplementation, options) {
    this.httpClient = httpClientImplementation;
    this.shouldTransformDates = options?.transformDates ?? true;
  }
  /**
   * Gets the specified settings object
   *
   * **Required scope:** app-settings:objects:read<br/>**Required permission:** app-settings:objects:read<br/><br/>Gets the specified settings object. Properties of type **secret** will be included in plain text if the call originates from a serverless function of your app; they will have irreversibly masked values otherwise. This protects these secrets from leaking to users of your app or other third parties.
   *
   * @deprecated
   *
   * @returns {Promise<AppSettingsObject>} Success
   *
   * @example <caption>Code example</caption>
   * import { appSettingsObjectsClient } from "@dynatrace-sdk/client-app-settings";
   *
   * const data =
   *   await appSettingsObjectsClient.getAppSettingsObjectByObjectId(
   *     { objectId: "..." },
   *   );
   *
   * @throws {AppSettingsErrorEnvelopeError} Failed. The input is invalid. | Failed. Forbidden. | No object available for the given objectId
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async getAppSettingsObjectByObjectId(config) {
    try {
      const response = await this.httpClient.send({
        url: `/platform/app-settings/v1/objects/${config.objectId}`,
        method: "GET",
        headers: {
          Accept: "application/json; charset=utf-8, application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? transformResponse(responseValue, [
          ["modificationInfo.createdTime", { datetime: true }],
          ["modificationInfo.lastModifiedTime", { datetime: true }]
        ]) : responseValue;
      } catch (err) {
        throw new InvalidResponseError(
          `AppSettingsObjectsClient.getAppSettingsObjectByObjectId:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if (isHttpClientAbortError(e) || isInvalidResponseError(e)) {
        throw e;
      }
      if (!isHttpClientResponseError(e)) {
        throw new ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AppSettingsErrorEnvelopeError(
              `400`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Failed. The input is invalid."),
              e
            );
          } catch (err) {
            if (err instanceof AppSettingsErrorEnvelopeError) {
              throw err;
            }
            throw new InvalidResponseError(
              `AppSettingsObjectsClient.getAppSettingsObjectByObjectId:400`,
              err,
              responseValue,
              "AppSettingsErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AppSettingsErrorEnvelopeError(
              `403`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Failed. Forbidden."),
              e
            );
          } catch (err) {
            if (err instanceof AppSettingsErrorEnvelopeError) {
              throw err;
            }
            throw new InvalidResponseError(
              `AppSettingsObjectsClient.getAppSettingsObjectByObjectId:403`,
              err,
              responseValue,
              "AppSettingsErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AppSettingsErrorEnvelopeError(
              `404`,
              response,
              errorBody,
              getErrorMessage(errorBody, "No object available for the given objectId"),
              e
            );
          } catch (err) {
            if (err instanceof AppSettingsErrorEnvelopeError) {
              throw err;
            }
            throw new InvalidResponseError(
              `AppSettingsObjectsClient.getAppSettingsObjectByObjectId:404`,
              err,
              responseValue,
              "AppSettingsErrorEnvelope",
              void 0,
              response
            );
          }
        }
        default: {
          try {
            const responseValue = await response.body("json");
            throw new ClientRequestError2(
              `${response.status}`,
              response,
              responseValue,
              getErrorMessage(
                responseValue,
                `Unexpected api response: code=${response.status} body="${responseValue}"`
              ),
              e
            );
          } catch (error) {
            if (error instanceof ClientRequestError2) {
              throw error;
            }
            const responseValue = await response.body("text").catch(() => "");
            throw new InvalidResponseError(
              `${response.status}`,
              error,
              responseValue,
              "json",
              "Unexpected api response",
              response
            );
          }
        }
      }
    }
  }
  /**
   * Updates an existing settings object
   *
   * **Required scope:** app-settings:objects:write<br/>**Required permission:** app-settings:objects:write<br/><br/>Updates an existing settings object with new values. To update a property of the `secret` type you need to pass the new value unmasked. To keep the current value, send the current masked secret. You can obtain it via [GET an object](https://dt-url.net/kj036oj) endpoint.
   *
   * Some schemas don't allow passing of the masked secret. In that case you need to send the unmasked secret with every update of the object.
   *
   * @deprecated
   *
   * @example <caption>Code example</caption>
   * import { appSettingsObjectsClient } from "@dynatrace-sdk/client-app-settings";
   *
   * const data =
   *   await appSettingsObjectsClient.putAppSettingsObjectByObjectId(
   *     {
   *       objectId: "...",
   *       optimisticLockingVersion: "...",
   *       body: { value: {} },
   *     },
   *   );
   *
   * @throws {AppSettingsErrorEnvelopeError} Failed. The input is invalid. | Failed. Forbidden. | Failed. The requested resource doesn't exist. | Failed. Conflicting resource.
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async putAppSettingsObjectByObjectId(config) {
    if (!config) {
      throw new ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const encodedBody = this.shouldTransformDates ? transformRequest(config.body, []) : config.body;
    const query = toQueryString({ "optimistic-locking-version": config.optimisticLockingVersion });
    try {
      await this.httpClient.send({
        url: `/platform/app-settings/v1/objects/${config.objectId}${query}`,
        method: "PUT",
        requestBodyType: "json",
        body: encodedBody,
        headers: {
          "Content-Type": "application/json; charset=utf-8"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      return;
    } catch (e) {
      if (isHttpClientAbortError(e)) {
        throw e;
      }
      if (!isHttpClientResponseError(e)) {
        throw new ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AppSettingsErrorEnvelopeError(
              `400`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Failed. The input is invalid."),
              e
            );
          } catch (err) {
            if (err instanceof AppSettingsErrorEnvelopeError) {
              throw err;
            }
            throw new InvalidResponseError(
              `AppSettingsObjectsClient.putAppSettingsObjectByObjectId:400`,
              err,
              responseValue,
              "AppSettingsErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AppSettingsErrorEnvelopeError(
              `403`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Failed. Forbidden."),
              e
            );
          } catch (err) {
            if (err instanceof AppSettingsErrorEnvelopeError) {
              throw err;
            }
            throw new InvalidResponseError(
              `AppSettingsObjectsClient.putAppSettingsObjectByObjectId:403`,
              err,
              responseValue,
              "AppSettingsErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AppSettingsErrorEnvelopeError(
              `404`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Failed. The requested resource doesn't exist."),
              e
            );
          } catch (err) {
            if (err instanceof AppSettingsErrorEnvelopeError) {
              throw err;
            }
            throw new InvalidResponseError(
              `AppSettingsObjectsClient.putAppSettingsObjectByObjectId:404`,
              err,
              responseValue,
              "AppSettingsErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 409: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AppSettingsErrorEnvelopeError(
              `409`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Failed. Conflicting resource."),
              e
            );
          } catch (err) {
            if (err instanceof AppSettingsErrorEnvelopeError) {
              throw err;
            }
            throw new InvalidResponseError(
              `AppSettingsObjectsClient.putAppSettingsObjectByObjectId:409`,
              err,
              responseValue,
              "AppSettingsErrorEnvelope",
              void 0,
              response
            );
          }
        }
        default: {
          try {
            const responseValue = await response.body("json");
            throw new ClientRequestError2(
              `${response.status}`,
              response,
              responseValue,
              getErrorMessage(
                responseValue,
                `Unexpected api response: code=${response.status} body="${responseValue}"`
              ),
              e
            );
          } catch (error) {
            if (error instanceof ClientRequestError2) {
              throw error;
            }
            const responseValue = await response.body("text").catch(() => "");
            throw new InvalidResponseError(
              `${response.status}`,
              error,
              responseValue,
              "json",
              "Unexpected api response",
              response
            );
          }
        }
      }
    }
  }
  /**
   * Deletes the specified settings object
   *
   * **Required scope:** app-settings:objects:write<br/>**Required permission:** app-settings:objects:write
   *
   * @deprecated
   *
   * @example <caption>Code example</caption>
   * import { appSettingsObjectsClient } from "@dynatrace-sdk/client-app-settings";
   *
   * const data =
   *   await appSettingsObjectsClient.deleteAppSettingsObjectByObjectId(
   *     { objectId: "...", optimisticLockingVersion: "..." },
   *   );
   *
   * @throws {AppSettingsErrorEnvelopeError} Failed. The input is invalid. | Failed. Forbidden. | Failed. The requested resource doesn't exist. | Failed. Conflicting resource.
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async deleteAppSettingsObjectByObjectId(config) {
    if (!config) {
      throw new ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const query = toQueryString({ "optimistic-locking-version": config.optimisticLockingVersion });
    try {
      await this.httpClient.send({
        url: `/platform/app-settings/v1/objects/${config.objectId}${query}`,
        method: "DELETE",
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [204].includes(status);
        }
      });
      return;
    } catch (e) {
      if (isHttpClientAbortError(e)) {
        throw e;
      }
      if (!isHttpClientResponseError(e)) {
        throw new ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AppSettingsErrorEnvelopeError(
              `400`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Failed. The input is invalid."),
              e
            );
          } catch (err) {
            if (err instanceof AppSettingsErrorEnvelopeError) {
              throw err;
            }
            throw new InvalidResponseError(
              `AppSettingsObjectsClient.deleteAppSettingsObjectByObjectId:400`,
              err,
              responseValue,
              "AppSettingsErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AppSettingsErrorEnvelopeError(
              `403`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Failed. Forbidden."),
              e
            );
          } catch (err) {
            if (err instanceof AppSettingsErrorEnvelopeError) {
              throw err;
            }
            throw new InvalidResponseError(
              `AppSettingsObjectsClient.deleteAppSettingsObjectByObjectId:403`,
              err,
              responseValue,
              "AppSettingsErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AppSettingsErrorEnvelopeError(
              `404`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Failed. The requested resource doesn't exist."),
              e
            );
          } catch (err) {
            if (err instanceof AppSettingsErrorEnvelopeError) {
              throw err;
            }
            throw new InvalidResponseError(
              `AppSettingsObjectsClient.deleteAppSettingsObjectByObjectId:404`,
              err,
              responseValue,
              "AppSettingsErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 409: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AppSettingsErrorEnvelopeError(
              `409`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Failed. Conflicting resource."),
              e
            );
          } catch (err) {
            if (err instanceof AppSettingsErrorEnvelopeError) {
              throw err;
            }
            throw new InvalidResponseError(
              `AppSettingsObjectsClient.deleteAppSettingsObjectByObjectId:409`,
              err,
              responseValue,
              "AppSettingsErrorEnvelope",
              void 0,
              response
            );
          }
        }
        default: {
          try {
            const responseValue = await response.body("json");
            throw new ClientRequestError2(
              `${response.status}`,
              response,
              responseValue,
              getErrorMessage(
                responseValue,
                `Unexpected api response: code=${response.status} body="${responseValue}"`
              ),
              e
            );
          } catch (error) {
            if (error instanceof ClientRequestError2) {
              throw error;
            }
            const responseValue = await response.body("text").catch(() => "");
            throw new InvalidResponseError(
              `${response.status}`,
              error,
              responseValue,
              "json",
              "Unexpected api response",
              response
            );
          }
        }
      }
    }
  }
  /**
   * Lists persisted settings objects
   *
   * **Required scope:** app-settings:objects:read<br/>**Required permission:** app-settings:objects:read<br/><br/>Lists persisted settings objects for selected schemas.
   *
   * If nothing is persisted or if all persisted settings objects are not accessible due to missing permissions, no items will be returned.
   *
   * To query the effective values (including schema defaults) please see [getEffectiveAppSettingsValues](#geteffectiveappsettingsvalues).
   *
   * Properties of type **secret** will be included in plain text if the call originates from a serverless function of your app; they will have irreversibly masked values otherwise. This protects these secrets from leaking to users of your app or other third parties.
   *
   * @deprecated
   *
   * @returns {Promise<AppSettingsObjectsList>} Success. Accessible objects returned.
   *
   *  Even if a response returns a successful response code it is possible that the result is incomplete due to an internal error.
   *
   *  In this case an 'error' property with information about the problem is added. The caller may decide to work with the incomplete result or do a retry of the operation.
   *
   * @example <caption>Code example</caption>
   * import { appSettingsObjectsClient } from "@dynatrace-sdk/client-app-settings";
   *
   * const data =
   *   await appSettingsObjectsClient.getAppSettingsObjects();
   *
   * @throws {AppSettingsErrorEnvelopeError} Failed. The input is invalid. | Failed. Forbidden. | Failed. The specified schema was not found.
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async getAppSettingsObjects(config = {}) {
    if (!config) {
      throw new ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const query = toQueryString({
      "schema-ids": config.schemaIds,
      "add-fields": config.addFields,
      "page-key": config.pageKey,
      "page-size": config.pageSize
    });
    try {
      const response = await this.httpClient.send({
        url: `/platform/app-settings/v1/objects${query}`,
        method: "GET",
        headers: {
          Accept: "application/json; charset=utf-8, application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? transformResponse(responseValue, [
          ["items.modificationInfo.createdTime", { datetime: true }],
          ["items.modificationInfo.lastModifiedTime", { datetime: true }]
        ]) : responseValue;
      } catch (err) {
        throw new InvalidResponseError(
          `AppSettingsObjectsClient.getAppSettingsObjects:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if (isHttpClientAbortError(e) || isInvalidResponseError(e)) {
        throw e;
      }
      if (!isHttpClientResponseError(e)) {
        throw new ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AppSettingsErrorEnvelopeError(
              `400`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Failed. The input is invalid."),
              e
            );
          } catch (err) {
            if (err instanceof AppSettingsErrorEnvelopeError) {
              throw err;
            }
            throw new InvalidResponseError(
              `AppSettingsObjectsClient.getAppSettingsObjects:400`,
              err,
              responseValue,
              "AppSettingsErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AppSettingsErrorEnvelopeError(
              `403`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Failed. Forbidden."),
              e
            );
          } catch (err) {
            if (err instanceof AppSettingsErrorEnvelopeError) {
              throw err;
            }
            throw new InvalidResponseError(
              `AppSettingsObjectsClient.getAppSettingsObjects:403`,
              err,
              responseValue,
              "AppSettingsErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AppSettingsErrorEnvelopeError(
              `404`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Failed. The specified schema was not found."),
              e
            );
          } catch (err) {
            if (err instanceof AppSettingsErrorEnvelopeError) {
              throw err;
            }
            throw new InvalidResponseError(
              `AppSettingsObjectsClient.getAppSettingsObjects:404`,
              err,
              responseValue,
              "AppSettingsErrorEnvelope",
              void 0,
              response
            );
          }
        }
        default: {
          try {
            const responseValue = await response.body("json");
            throw new ClientRequestError2(
              `${response.status}`,
              response,
              responseValue,
              getErrorMessage(
                responseValue,
                `Unexpected api response: code=${response.status} body="${responseValue}"`
              ),
              e
            );
          } catch (error) {
            if (error instanceof ClientRequestError2) {
              throw error;
            }
            const responseValue = await response.body("text").catch(() => "");
            throw new InvalidResponseError(
              `${response.status}`,
              error,
              responseValue,
              "json",
              "Unexpected api response",
              response
            );
          }
        }
      }
    }
  }
  /**
   * Creates a new settings object
   *
   * **Required scope:** app-settings:objects:write<br/>**Required permission:** app-settings:objects:write<br/><br/>Creates a new settings object.
   *
   * @deprecated
   *
   * @returns {Promise<AppSettingsObjectResponse | void>} Created
   *
   * @example <caption>Code example</caption>
   * import { appSettingsObjectsClient } from "@dynatrace-sdk/client-app-settings";
   *
   * const data =
   *   await appSettingsObjectsClient.postAppSettingsObject({
   *     body: { schemaId: "jira-connection", value: {} },
   *   });
   *
   * @throws {AppSettingsErrorEnvelopeError} Failed. The input is invalid. | Failed. Forbidden. | Failed. The requested resource doesn't exist. | Failed. Conflicting resource.
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async postAppSettingsObject(config) {
    if (!config) {
      throw new ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const encodedBody = this.shouldTransformDates ? transformRequest(config.body, []) : config.body;
    const query = toQueryString({ "validate-only": config.validateOnly });
    try {
      const response = await this.httpClient.send({
        url: `/platform/app-settings/v1/objects${query}`,
        method: "POST",
        requestBodyType: "json",
        body: encodedBody,
        headers: {
          "Content-Type": "application/json; charset=utf-8",
          Accept: "application/json; charset=utf-8, application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200, 201].includes(status);
        }
      });
      switch (response.status) {
        case 200: {
          return;
        }
        case 201: {
          const responseValue = await response.body("json");
          try {
            return this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
          } catch (err) {
            throw new InvalidResponseError(
              `AppSettingsObjectsClient.postAppSettingsObject:${response.status}`,
              err,
              responseValue,
              void 0,
              void 0,
              void 0
            );
          }
        }
        default: {
          try {
            const responseValue = await response.body("json");
            throw new ClientRequestError2(
              `${response.status}`,
              response,
              responseValue,
              getErrorMessage(
                responseValue,
                `Unexpected api response: code=${response.status} body="${responseValue}"`
              )
            );
          } catch (error) {
            if (error instanceof ClientRequestError2) {
              throw error;
            }
            const responseValue = await response.body("text").catch(() => "");
            throw new InvalidResponseError(
              `${response.status}`,
              error,
              responseValue,
              "json",
              "Unexpected api response",
              response
            );
          }
        }
      }
    } catch (e) {
      if (isHttpClientAbortError(e) || isInvalidResponseError(e) || isClientRequestError(e)) {
        throw e;
      }
      if (!isHttpClientResponseError(e)) {
        throw new ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AppSettingsErrorEnvelopeError(
              `400`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Failed. The input is invalid."),
              e
            );
          } catch (err) {
            if (err instanceof AppSettingsErrorEnvelopeError) {
              throw err;
            }
            throw new InvalidResponseError(
              `AppSettingsObjectsClient.postAppSettingsObject:400`,
              err,
              responseValue,
              "AppSettingsErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AppSettingsErrorEnvelopeError(
              `403`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Failed. Forbidden."),
              e
            );
          } catch (err) {
            if (err instanceof AppSettingsErrorEnvelopeError) {
              throw err;
            }
            throw new InvalidResponseError(
              `AppSettingsObjectsClient.postAppSettingsObject:403`,
              err,
              responseValue,
              "AppSettingsErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AppSettingsErrorEnvelopeError(
              `404`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Failed. The requested resource doesn't exist."),
              e
            );
          } catch (err) {
            if (err instanceof AppSettingsErrorEnvelopeError) {
              throw err;
            }
            throw new InvalidResponseError(
              `AppSettingsObjectsClient.postAppSettingsObject:404`,
              err,
              responseValue,
              "AppSettingsErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 409: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AppSettingsErrorEnvelopeError(
              `409`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Failed. Conflicting resource."),
              e
            );
          } catch (err) {
            if (err instanceof AppSettingsErrorEnvelopeError) {
              throw err;
            }
            throw new InvalidResponseError(
              `AppSettingsObjectsClient.postAppSettingsObject:409`,
              err,
              responseValue,
              "AppSettingsErrorEnvelope",
              void 0,
              response
            );
          }
        }
        default: {
          try {
            const responseValue = await response.body("json");
            throw new ClientRequestError2(
              `${response.status}`,
              response,
              responseValue,
              getErrorMessage(
                responseValue,
                `Unexpected api response: code=${response.status} body="${responseValue}"`
              ),
              e
            );
          } catch (error) {
            if (error instanceof ClientRequestError2) {
              throw error;
            }
            const responseValue = await response.body("text").catch(() => "");
            throw new InvalidResponseError(
              `${response.status}`,
              error,
              responseValue,
              "json",
              "Unexpected api response",
              response
            );
          }
        }
      }
    }
  }
  /**
   * Lists effective settings values
   *
   * **Required scope:** app-settings:objects:read<br/>**Required permission:** app-settings:objects:read<br/><br/>Lists effective settings values for selected schemas.
   * If no object is persisted for a schema with **"multiObject": false**, the default value as defined in the schema will be returned.
   *
   * Properties of type **secret** will be included in plain text if the call originates from a serverless function of your app; they will have irreversibly masked values otherwise. This protects these secrets from leaking to users of your app or other third parties.
   *
   * @deprecated
   *
   * @returns {Promise<EffectiveAppSettingsValuesList>} Success
   *
   *  Even if a response returns a successful response code it is possible that the result is incomplete due to an internal error.
   *
   *  In this case an 'error' property with information about the problem is added. The caller may decide to work with the incomplete result or do a retry of the operation.
   *
   * @example <caption>Code example</caption>
   * import { appSettingsObjectsClient } from "@dynatrace-sdk/client-app-settings";
   *
   * const data =
   *   await appSettingsObjectsClient.getEffectiveAppSettingsValues();
   *
   * @throws {AppSettingsErrorEnvelopeError} Failed. The input is invalid. | Failed. The specified schema is not found.
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async getEffectiveAppSettingsValues(config = {}) {
    if (!config) {
      throw new ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const query = toQueryString({
      "schema-ids": config.schemaIds,
      "add-fields": config.addFields,
      "page-key": config.pageKey,
      "page-size": config.pageSize
    });
    try {
      const response = await this.httpClient.send({
        url: `/platform/app-settings/v1/effective-values${query}`,
        method: "GET",
        headers: {
          Accept: "application/json; charset=utf-8, application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? transformResponse(responseValue, [
          ["items.modificationInfo.createdTime", { datetime: true }],
          ["items.modificationInfo.lastModifiedTime", { datetime: true }]
        ]) : responseValue;
      } catch (err) {
        throw new InvalidResponseError(
          `AppSettingsObjectsClient.getEffectiveAppSettingsValues:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if (isHttpClientAbortError(e) || isInvalidResponseError(e)) {
        throw e;
      }
      if (!isHttpClientResponseError(e)) {
        throw new ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AppSettingsErrorEnvelopeError(
              `400`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Failed. The input is invalid."),
              e
            );
          } catch (err) {
            if (err instanceof AppSettingsErrorEnvelopeError) {
              throw err;
            }
            throw new InvalidResponseError(
              `AppSettingsObjectsClient.getEffectiveAppSettingsValues:400`,
              err,
              responseValue,
              "AppSettingsErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AppSettingsErrorEnvelopeError(
              `404`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Failed. The specified schema is not found."),
              e
            );
          } catch (err) {
            if (err instanceof AppSettingsErrorEnvelopeError) {
              throw err;
            }
            throw new InvalidResponseError(
              `AppSettingsObjectsClient.getEffectiveAppSettingsValues:404`,
              err,
              responseValue,
              "AppSettingsErrorEnvelope",
              void 0,
              response
            );
          }
        }
        default: {
          try {
            const responseValue = await response.body("json");
            throw new ClientRequestError2(
              `${response.status}`,
              response,
              responseValue,
              getErrorMessage(
                responseValue,
                `Unexpected api response: code=${response.status} body="${responseValue}"`
              ),
              e
            );
          } catch (error) {
            if (error instanceof ClientRequestError2) {
              throw error;
            }
            const responseValue = await response.body("text").catch(() => "");
            throw new InvalidResponseError(
              `${response.status}`,
              error,
              responseValue,
              "json",
              "Unexpected api response",
              response
            );
          }
        }
      }
    }
  }
  /**
   * Get the effective settings permissions of the calling user in the environment
   *
   * **Required scope:** app-settings:objects:read or app-settings:objects:write<br/>**Required permission:** app-settings:objects:read or app-settings:objects:write
   *
   * @deprecated
   *
   * @returns {Promise<EffectivePermissions>} Success
   *
   * @example <caption>Code example</caption>
   * import { appSettingsObjectsClient } from "@dynatrace-sdk/client-app-settings";
   *
   * const data =
   *   await appSettingsObjectsClient.resolveEffectivePermissions(
   *     {
   *       body: {
   *         permissions: [
   *           {
   *             permission: "app-settings:objects:read",
   *             context: { schemaId: "..." },
   *           },
   *         ],
   *       },
   *     },
   *   );
   *
   * @throws {AppSettingsErrorEnvelopeError} Failed. The input is invalid. | Failed. Forbidden. No access to any schema. | Failed.
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async resolveEffectivePermissions(config) {
    if (!config) {
      throw new ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const encodedBody = this.shouldTransformDates ? transformRequest(config.body, []) : config.body;
    try {
      const response = await this.httpClient.send({
        url: `/platform/app-settings/v1/effective-permissions:resolve`,
        method: "POST",
        requestBodyType: "json",
        body: encodedBody,
        headers: {
          "Content-Type": "application/json; charset=utf-8",
          Accept: "application/json; charset=utf-8, application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
      } catch (err) {
        throw new InvalidResponseError(
          `AppSettingsObjectsClient.resolveEffectivePermissions:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if (isHttpClientAbortError(e) || isInvalidResponseError(e)) {
        throw e;
      }
      if (!isHttpClientResponseError(e)) {
        throw new ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AppSettingsErrorEnvelopeError(
              `400`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Failed. The input is invalid."),
              e
            );
          } catch (err) {
            if (err instanceof AppSettingsErrorEnvelopeError) {
              throw err;
            }
            throw new InvalidResponseError(
              `AppSettingsObjectsClient.resolveEffectivePermissions:400`,
              err,
              responseValue,
              "AppSettingsErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AppSettingsErrorEnvelopeError(
              `403`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Failed. Forbidden. No access to any schema."),
              e
            );
          } catch (err) {
            if (err instanceof AppSettingsErrorEnvelopeError) {
              throw err;
            }
            throw new InvalidResponseError(
              `AppSettingsObjectsClient.resolveEffectivePermissions:403`,
              err,
              responseValue,
              "AppSettingsErrorEnvelope",
              void 0,
              response
            );
          }
        }
        default: {
          if (response.status >= 400 && response.status <= 499) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
              throw new AppSettingsErrorEnvelopeError(
                `${response.status}`,
                response,
                errorBody,
                getErrorMessage(errorBody, "Failed."),
                e
              );
            } catch (err) {
              if (err instanceof AppSettingsErrorEnvelopeError) {
                throw err;
              }
              throw new InvalidResponseError(
                `AppSettingsObjectsClient.resolveEffectivePermissions:${response.status}`,
                err,
                responseValue,
                "AppSettingsErrorEnvelope",
                void 0,
                response
              );
            }
          } else if (response.status >= 500 && response.status <= 599) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
              throw new AppSettingsErrorEnvelopeError(
                `${response.status}`,
                response,
                errorBody,
                getErrorMessage(errorBody, "Failed."),
                e
              );
            } catch (err) {
              if (err instanceof AppSettingsErrorEnvelopeError) {
                throw err;
              }
              throw new InvalidResponseError(
                `AppSettingsObjectsClient.resolveEffectivePermissions:${response.status}`,
                err,
                responseValue,
                "AppSettingsErrorEnvelope",
                void 0,
                response
              );
            }
          } else {
            try {
              const responseValue = await response.body("json");
              throw new ClientRequestError2(
                `${response.status}`,
                response,
                responseValue,
                getErrorMessage(
                  responseValue,
                  `Unexpected api response: code=${response.status} body="${responseValue}"`
                )
              );
            } catch (error) {
              if (error instanceof ClientRequestError2) {
                throw error;
              }
              const responseValue = await response.body("text").catch(() => "");
              throw new InvalidResponseError(
                `${response.status}`,
                error,
                responseValue,
                "json",
                "Unexpected api response",
                response
              );
            }
          }
        }
      }
    }
  }
};
var appSettingsObjectsClient = /* @__PURE__ */ new AppSettingsObjectsClient(defaultHttpClient);

// packages/client/app-settings/src/lib/error-envelopes/index.ts
import {
  ApiClientError as ApiClientError2,
  ApiGatewayError as ApiGatewayError2,
  ClientRequestError as ClientRequestError3,
  InvalidResponseError as InvalidResponseError2,
  isApiClientError,
  isApiGatewayError,
  isClientRequestError as isClientRequestError2,
  isInvalidResponseError as isInvalidResponseError2
} from "@dynatrace-sdk/shared-errors";

// packages/client/app-settings/src/lib/models/constraint-violation-parameter-location.ts
var ConstraintViolationParameterLocation = /* @__PURE__ */ ((ConstraintViolationParameterLocation2) => {
  ConstraintViolationParameterLocation2["Header"] = "HEADER";
  ConstraintViolationParameterLocation2["Path"] = "PATH";
  ConstraintViolationParameterLocation2["PayloadBody"] = "PAYLOAD_BODY";
  ConstraintViolationParameterLocation2["Query"] = "QUERY";
  return ConstraintViolationParameterLocation2;
})(ConstraintViolationParameterLocation || {});

// packages/client/app-settings/src/lib/models/effective-permission-granted.ts
var EffectivePermissionGranted = /* @__PURE__ */ ((EffectivePermissionGranted2) => {
  EffectivePermissionGranted2["True"] = "true";
  EffectivePermissionGranted2["False"] = "false";
  EffectivePermissionGranted2["Condition"] = "condition";
  return EffectivePermissionGranted2;
})(EffectivePermissionGranted || {});

// packages/client/app-settings/src/lib/models/resource-context-operations-item.ts
var ResourceContextOperationsItem = /* @__PURE__ */ ((ResourceContextOperationsItem2) => {
  ResourceContextOperationsItem2["Read"] = "read";
  ResourceContextOperationsItem2["Write"] = "write";
  ResourceContextOperationsItem2["Delete"] = "delete";
  return ResourceContextOperationsItem2;
})(ResourceContextOperationsItem || {});

// packages/client/app-settings/src/lib/models/single-permission-request-permission.ts
var SinglePermissionRequestPermission = /* @__PURE__ */ ((SinglePermissionRequestPermission2) => {
  SinglePermissionRequestPermission2["AppSettingsObjectsRead"] = "app-settings:objects:read";
  SinglePermissionRequestPermission2["AppSettingsObjectsWrite"] = "app-settings:objects:write";
  return SinglePermissionRequestPermission2;
})(SinglePermissionRequestPermission || {});
export {
  ApiClientError2 as ApiClientError,
  ApiGatewayError2 as ApiGatewayError,
  AppSettingsErrorEnvelopeError,
  AppSettingsObjectsClient,
  ClientRequestError3 as ClientRequestError,
  ConstraintViolationParameterLocation,
  EffectivePermissionGranted,
  InvalidResponseError2 as InvalidResponseError,
  ResourceContextOperationsItem,
  SinglePermissionRequestPermission,
  appSettingsObjectsClient,
  isApiClientError,
  isApiGatewayError,
  isAppSettingsErrorEnvelopeError,
  isClientRequestError2 as isClientRequestError,
  isInvalidResponseError2 as isInvalidResponseError
};
