// @dynatrace/strato-components/charts/meter-bar/single-meter-bar/types/default.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var default_exports = {};
__export(default_exports, {
  NAME_DEFAULT_VALUE: () => NAME_DEFAULT_VALUE,
  SEGMENT_COLOR_DEFAULT: () => SEGMENT_COLOR_DEFAULT,
  THRESHOLD_COLOR_DEFAULT: () => THRESHOLD_COLOR_DEFAULT,
  THRESHOLD_METADATA_DEFAULT: () => THRESHOLD_METADATA_DEFAULT,
  defaultConfig: () => defaultConfig
});
module.exports = __toCommonJS(default_exports);
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
var import_constants = require("../../shared/constants.js");
const NAME_DEFAULT_VALUE = "value";
const SEGMENT_COLOR_DEFAULT = import_colors.default.Charts.Categorical.Color11.Default;
const THRESHOLD_COLOR_DEFAULT = SEGMENT_COLOR_DEFAULT;
const THRESHOLD_METADATA_DEFAULT = {
  data: [],
  legend: false
};
const defaultConfig = {
  loading: false,
  thresholdIndicators: THRESHOLD_METADATA_DEFAULT,
  size: import_constants.SIZE_DEFAULT,
  min: import_constants.MIN_DEFAULT_VALUE,
  max: import_constants.MAX_DEFAULT_VALUE,
  formatter: import_constants.DEFAULT_FORMATTER,
  tooltip: { hidden: false }
};
