// @dynatrace-sdk/shared-client-utils/cjs/index.js
/**
 * @license
 * Copyright 2023-2025 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/shared-client-utils/src/index.ts
var src_exports = {};
__export(src_exports, {
  Encoding: () => Encoding,
  apiGatewayErrorsHandler: () => apiGatewayErrorsHandler,
  getErrorMessage: () => getErrorMessage,
  mimeTypeToEncoding: () => mimeTypeToEncoding,
  toQueryString: () => toQueryString,
  transformFromFormData: () => transformFromFormData,
  transformRequest: () => transformRequest,
  transformResponse: () => transformResponse,
  transformToFormData: () => transformToFormData,
  transformToFormDataField: () => transformToFormDataField
});
module.exports = __toCommonJS(src_exports);

// packages/shared-client-utils/src/api-gateway-errors-handler.ts
var import_shared_errors = require("@dynatrace-sdk/shared-errors");
async function apiGatewayErrorsHandler(response) {
  if (response.headers?.["dynatrace-response-source"] === "API Gateway") {
    const responseBody = await response.body("json");
    throw new import_shared_errors.ApiGatewayError(response, responseBody);
  }
}

// packages/shared-client-utils/src/encoding.ts
var Encoding = /* @__PURE__ */ ((Encoding2) => {
  Encoding2["Json"] = "json";
  Encoding2["Text"] = "text";
  Encoding2["Binary"] = "binary";
  Encoding2["FormData"] = "form-data";
  return Encoding2;
})(Encoding || {});
function mimeTypeToEncoding(mimeType, defaultEncoding) {
  const shortMimeType = mimeType.split(";")[0];
  switch (shortMimeType) {
    case "application/json":
    case "application/cloudevent+json":
    case "application/cloudevent-batch+json":
      return "json" /* Json */;
    case "text/plain":
      return "text" /* Text */;
    case "application/octet-stream":
      return "binary" /* Binary */;
    case "multipart/form-data":
      return "form-data" /* FormData */;
    default:
      if (defaultEncoding)
        return defaultEncoding;
      else
        throw new Error(`${shortMimeType} mime type is not yet supported`);
  }
}

// packages/shared-client-utils/src/get-error-message.ts
function serializeData(data) {
  try {
    return JSON.stringify(data);
  } catch (e) {
    return String(data);
  }
}
function getMessagesFromErrorDetails(details) {
  const messages = [];
  Object.entries(details).forEach(([name, data]) => {
    if (!data)
      return;
    const serializedData = serializeData(data);
    switch (name) {
      case "missingScopes":
        messages.push(`Missing scopes: ${serializedData}`);
        break;
      default:
        messages.push(`${name}: ${serializedData}`);
    }
  }, []);
  return messages;
}
function getErrorMessage(errorBody, defaultMessage) {
  const error = errorBody;
  const msg = error?.error?.message || defaultMessage;
  const details = error?.error?.details || {};
  return [msg, ...getMessagesFromErrorDetails(details)].join(". ");
}

// packages/shared-client-utils/src/transformations.ts
var defaultRuleConfig = {
  force: false,
  datetime: false
};
function mergeKeys(key1, key2) {
  if (!key1)
    return key2;
  if (!key2)
    return key1;
  return `${key1}.${key2}`;
}
function matchKey(key, rules, cache) {
  if (cache.has(key)) {
    return cache.get(key);
  }
  for (const rule of rules) {
    const pattern = Array.isArray(rule) ? rule[0] : rule;
    const config = Array.isArray(rule) ? rule[1] : defaultRuleConfig;
    if (key === pattern) {
      const result = [pattern, config];
      cache.set(key, result);
      return result;
    }
    let i = 0;
    let j = 0;
    const keyLen = key.length;
    const patternLen = pattern.length;
    let isMatch = true;
    while (i < keyLen && j < patternLen) {
      let keyPartEnd = key.indexOf(".", i);
      let patternPartEnd = pattern.indexOf(".", j);
      if (keyPartEnd === -1)
        keyPartEnd = keyLen;
      if (patternPartEnd === -1)
        patternPartEnd = patternLen;
      const keyPart = key.slice(i, keyPartEnd);
      const patternPart = pattern.slice(j, patternPartEnd);
      if (patternPart !== "*" && patternPart !== "**" && keyPart !== patternPart) {
        isMatch = false;
        break;
      }
      i = keyPartEnd + 1;
      j = patternPartEnd + 1;
    }
    if (isMatch) {
      if (j < patternLen) {
        const remaining = pattern.slice(j);
        if (remaining !== "**" && remaining !== "*") {
          isMatch = false;
        }
      }
      if (i < keyLen && pattern.slice(j - 1) !== "**") {
        isMatch = false;
      }
    }
    if (isMatch) {
      const result = [pattern, config];
      cache.set(key, result);
      return result;
    }
  }
  cache.set(key, null);
  return null;
}
var potentialMatchForRoot = (pattern, key) => pattern.startsWith(key) || pattern.startsWith("*") || pattern.startsWith("**");
var potentialMatch = (pattern, key) => pattern.startsWith(key) || pattern.includes("*") || pattern.includes("**");
function partialMatch(key, keyPatterns) {
  const keyPatternsLen = keyPatterns.length;
  if (!key.includes(".")) {
    for (let i = 0; i < keyPatternsLen; i++) {
      const pattern = keyPatterns[i];
      if (potentialMatchForRoot(pattern, key))
        return true;
    }
    return false;
  }
  let dotIndex = -1;
  let currentKeyEnd = 0;
  while ((dotIndex = key.indexOf(".", currentKeyEnd)) !== -1) {
    const currentKey = key.slice(0, dotIndex);
    const isRoot = !currentKey.includes(".");
    for (let i = 0; i < keyPatternsLen; i++) {
      const pattern = keyPatterns[i];
      if (isRoot) {
        if (potentialMatchForRoot(pattern, currentKey))
          return true;
      } else if (potentialMatch(pattern, currentKey))
        return true;
    }
    currentKeyEnd = dotIndex + 1;
  }
  for (let i = 0; i < keyPatternsLen; i++) {
    const pattern = keyPatterns[i];
    if (potentialMatch(pattern, key))
      return true;
  }
  return false;
}
var dateTimePattern = /^(\d{4}-\d{2}-\d{2})(T\d{2}:\d{2}:\d{2}(?:\.\d+)?(Z|[+-]\d{2}:\d{2})?)?$/i;
function isDateLike(value) {
  if (typeof value === "string") {
    return dateTimePattern.test(value);
  }
  return false;
}
function transformValue(value, rules, direction) {
  const { force, datetime } = rules;
  if (direction === "to") {
    return force || isDateLike(value) ? new Date(value) : value;
  }
  if (value instanceof Date) {
    const iso = value.toISOString();
    return datetime ? iso : iso.split("T")[0];
  }
  return value;
}
function walk(obj, direction, rules, rulesCache, keyPatterns, currentKey = "") {
  if (rules.length <= 0) {
    return obj;
  }
  if (!partialMatch(currentKey, keyPatterns)) {
    return obj;
  }
  if (Array.isArray(obj)) {
    const match = matchKey(currentKey, rules, rulesCache);
    obj.forEach((item, idx) => {
      if (typeof item === "object" && item !== null) {
        walk(item, direction, rules, rulesCache, keyPatterns, currentKey);
      } else if (match) {
        obj[idx] = transformValue(obj[idx], match[1], direction);
      }
    });
  }
  if (typeof obj === "object" && obj !== null && !Array.isArray(obj)) {
    for (const key of Object.keys(obj)) {
      const mergedKey = mergeKeys(currentKey, key);
      const match = matchKey(mergedKey, rules, rulesCache);
      if (match) {
        obj[key] = transformValue(obj[key], match[1], direction);
      }
      walk(obj[key], direction, rules, rulesCache, keyPatterns, mergedKey);
    }
  }
  return obj;
}
function flattenKeyPatterns(keyPatterns) {
  return keyPatterns.map((k) => Array.isArray(k) ? k[0] : k);
}
function transform(direction, object, rules) {
  const flatKeyPatterns = flattenKeyPatterns(rules);
  const rulesCache = /* @__PURE__ */ new Map();
  return walk(object, direction, rules, rulesCache, flatKeyPatterns);
}
function transformRequest(object, rules) {
  return transform("from", object, rules);
}
function transformResponse(object, rules) {
  return transform("to", object, rules);
}
function transformToFormData(value, fields) {
  return fields.map((field) => transformToFormDataField(value[field.key], field)).filter((field) => Boolean(field));
}
function transformToFormDataField(value, field) {
  if (!field.required && value === void 0) {
    return null;
  }
  return {
    name: field.key,
    type: field.type,
    value: field.type === "text" ? String(value) : value
  };
}
function transformFromFormData(formData, fields) {
  return Object.fromEntries(
    formData.map(({ name, value }) => [
      name,
      fields[name]?.type === "json" ? transformResponse(
        JSON.parse(value),
        fields[name].transformations
      ) : value
    ])
  );
}

// packages/shared-client-utils/src/url-helpers.ts
var encodeQueryParam = (key, value) => {
  const encodedKey = encodeURIComponent(key);
  return `${encodedKey}=${encodeURIComponent(typeof value === "number" ? value : String(value))}`;
};
var addExplodedArrayQueryParam = (query, key) => {
  const arrayValue = query[key];
  return arrayValue.map((value) => encodeQueryParam(key, value)).join("&");
};
var addNonExplodedArrayQueryParams = (query, key) => {
  const encodedKey = encodeURIComponent(key);
  const encodedParamsList = query[key].map((value) => encodeURIComponent(typeof value === "number" ? value : String(value))).join(",");
  return `${encodedKey}=${encodedParamsList}`;
};
var addQueryParam = (query, key) => encodeQueryParam(key, query[key]);
var arrayQueryParams = (query, key, explode) => {
  return explode ? addExplodedArrayQueryParam(query, key) : addNonExplodedArrayQueryParams(query, key);
};
var toQueryString = (rawQuery, flags = {}) => {
  const query = rawQuery || {};
  const keys = Object.keys(query).filter((key) => typeof query[key] !== "undefined");
  const queryString = keys.map(
    (key) => Array.isArray(query[key]) ? arrayQueryParams(
      query,
      key,
      flags.explode?.hasOwnProperty(key) ? flags.explode[key] : true
    ) : addQueryParam(query, key)
  ).join("&");
  return queryString ? `?${queryString}` : "";
};
