// @dynatrace/strato-components/charts/categorical-bar/hooks/store/selectors.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var selectors_exports = {};
__export(selectors_exports, {
  useFilteredDimensionsState: () => useFilteredDimensionsState,
  useHighlightedDimensionState: () => useHighlightedDimensionState,
  useHoveredThresholdIdState: () => useHoveredThresholdIdState,
  useMouseInBoundsState: () => useMouseInBoundsState,
  useMousePositionState: () => useMousePositionState,
  useSelector: () => useSelector,
  useSetState: () => useSetState,
  useTooltipState: () => useTooltipState
});
module.exports = __toCommonJS(selectors_exports);
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_selector = require("../../../../core/store/selector.js");
var import_CategoricalBarChartState_context = require("../../contexts/CategoricalBarChartState.context.js");
const useSelector = (selector, isEqual2) => (0, import_selector.useStoreSelector)(import_CategoricalBarChartState_context.CategoricalBarChartStateContext, selector, isEqual2);
const useHighlightedDimensionState = () => useSelector((state) => state.highlightedDimension);
const useFilteredDimensionsState = () => useSelector((state) => state.filteredDimensions);
const useTooltipState = () => useSelector(
  (state) => ({
    pinnedTooltip: state.overlays.pinnedTooltip,
    visibleTooltip: state.overlays.visibleTooltip
  }),
  import_lodash_es.isEqual
);
const useHoveredThresholdIdState = () => useSelector((state) => state.hoveredThresholdId);
const useSetState = () => {
  const { setState } = (0, import_react.useContext)(import_CategoricalBarChartState_context.CategoricalBarChartStateContext);
  return setState;
};
const useMousePositionState = () => {
  const ref = (0, import_react.useRef)(void 0);
  return useSelector((state) => {
    if (!state.overlays.pinnedTooltip) {
      ref.current = state.overlays.mousePosition;
    }
    return ref.current;
  }, import_lodash_es.isEqual);
};
const useMouseInBoundsState = () => useSelector(
  (state) => state.overlays.mouseInBounds
);
