// @dynatrace-sdk/segments-presets/testing/esm/index.js
/**
 * @license
 * Copyright 2023-2025 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// packages/platform/segments-presets/testing/test-utils.ts
var currentSetup;
var isInitialised = false;
var wrapper = {
  getSegmentsPresets: () => currentSetup.getSegmentsPresets(),
  addRecentSegmentsPreset: (segments) => currentSetup.addRecentSegmentsPreset(segments),
  pinSegmentsPreset: (segments) => currentSetup.pinSegmentsPreset(segments),
  unpinSegmentsPreset: (id) => currentSetup.unpinSegmentsPreset(id)
};
var setupTestingEnv = () => {
  if (typeof expect === "undefined") {
    isInitialised = true;
  }
  mockSegmentsPresets();
};
var mockSegmentsPresets = (setup = {}) => {
  if (!isInitialised) {
    throw new Error(`\`@dynatrace-sdk/segments-presets/testing\` has not been initialised!
Add in your \`jest.config.js\` file entry with \`@dynatrace-sdk/segments-presets/testing\` setup:

\`setupFiles: ['@dynatrace-sdk/segments-presets/testing']\``);
  }
  const globalWithDtRuntimeSegments = globalThis;
  if (globalWithDtRuntimeSegments.dtRuntime === void 0) {
    globalWithDtRuntimeSegments.dtRuntime = {};
  }
  if (globalWithDtRuntimeSegments.dtRuntime.segments === void 0) {
    globalWithDtRuntimeSegments.dtRuntime.segments = wrapper;
  }
  currentSetup = getSetupWithDefaults(setup);
};
var clearSegmentsPresetsMocks = () => {
  mockSegmentsPresets();
};
var getSetupWithDefaults = (setup) => {
  return {
    getSegmentsPresets: async () => setup.getSegmentsPresets?.() || { pinnedPresets: [], recentPresets: [] },
    addRecentSegmentsPreset: async (segments) => setup.addRecentSegmentsPreset?.(segments) || { pinnedPresets: [], recentPresets: [] },
    pinSegmentsPreset: async (segments) => setup.pinSegmentsPreset?.(segments) || { pinnedPresets: [], recentPresets: [] },
    unpinSegmentsPreset: async (id) => setup.unpinSegmentsPreset?.(id) || { pinnedPresets: [], recentPresets: [] }
  };
};

// packages/platform/segments-presets/testing/index.ts
setupTestingEnv();
export {
  clearSegmentsPresetsMocks,
  mockSegmentsPresets,
  setupTestingEnv
};
