// @dynatrace-sdk/navigation/esm/index.js
/**
 * @license
 * Copyright 2023-2025 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// packages/platform/shared/utils/get-global-with-dt-runtime.ts
var getGlobalWithDtRuntime = () => typeof globalThis !== "undefined" ? globalThis : window;

// packages/platform/navigation/src/lib/get-intent.ts
function getIntent() {
  const getIntentFunc = getGlobalWithDtRuntime().dtRuntime?.navigation?.getIntent;
  if (typeof getIntentFunc === "function") {
    return getIntentFunc();
  } else {
    console.warn("Missing getIntent function from web runtime.");
    return null;
  }
}

// packages/platform/navigation/src/lib/get-intent-link.ts
function getIntentLink(intentPayload, appId, intentId) {
  const getIntentLinkFunc = getGlobalWithDtRuntime().dtRuntime?.navigation?.getIntentLink;
  if (typeof getIntentLinkFunc === "function") {
    return getIntentLinkFunc(intentPayload, appId, intentId);
  } else {
    console.warn("Missing getIntentLink function from web runtime.");
    return "https://dynatrace.com/";
  }
}

// packages/platform/shared/utils/types-guards.ts
function isString(val) {
  return typeof val === "string";
}

// packages/platform/navigation/src/lib/send-intent.ts
function sendIntent(intentPayload, legacyRecommendedAppIdOrSendIntentOptions, recommendedIntentId) {
  const sendIntentFunc = getGlobalWithDtRuntime().dtRuntime?.navigation?.sendIntent;
  if (typeof sendIntentFunc === "function") {
    if (isString(legacyRecommendedAppIdOrSendIntentOptions) && isString(recommendedIntentId)) {
      sendIntentFunc(intentPayload, legacyRecommendedAppIdOrSendIntentOptions, recommendedIntentId);
    } else if (legacyRecommendedAppIdOrSendIntentOptions !== void 0) {
      const sendIntentOptions = legacyRecommendedAppIdOrSendIntentOptions;
      sendIntentFunc(intentPayload, sendIntentOptions);
    } else {
      sendIntentFunc(intentPayload);
    }
  } else {
    console.warn("Missing sendIntent function from web runtime.");
  }
}

// packages/platform/navigation/src/lib/send-intent-with-response.ts
async function sendIntentWithResponse(intentPayload, sendIntentWithResponseOptions) {
  const sendIntentFunc = getGlobalWithDtRuntime().dtRuntime?.navigation?.sendIntentWithResponseV2;
  if (typeof sendIntentFunc === "function") {
    return sendIntentFunc(intentPayload, sendIntentWithResponseOptions);
  } else {
    console.warn("Missing sendIntentWithResponseOptions function from web runtime.");
    return void 0;
  }
}

// packages/platform/navigation/src/lib/open-app.ts
function openApp(appId, pageToken) {
  const openAppFunc = getGlobalWithDtRuntime().dtRuntime?.navigation?.openApp;
  if (typeof openAppFunc === "function") {
    openAppFunc(appId, pageToken);
  } else {
    console.warn("Missing openApp function from web runtime.");
  }
}

// packages/platform/navigation/src/lib/get-app-link.ts
function getAppLink(appId, pageToken) {
  const getAppLinkFunc = getGlobalWithDtRuntime().dtRuntime?.navigation?.getAppLink;
  if (typeof getAppLinkFunc === "function") {
    return getAppLinkFunc(appId, pageToken);
  } else {
    console.warn("Missing getAppLink function from web runtime.");
    return "https://dynatrace.com/";
  }
}

// packages/platform/navigation/src/lib/set-path-change-handler.ts
function setPathChangeHandler(handler) {
  const setPathChangeHandlerFunc = getGlobalWithDtRuntime().dtRuntime?.navigation?.setPathChangeHandler;
  if (typeof setPathChangeHandlerFunc === "function") {
    setPathChangeHandlerFunc(handler);
  } else {
    console.warn("Missing setPathChangeHandler function from web runtime.");
  }
}

// packages/platform/navigation/src/lib/get-document-link.ts
function getDocumentLink(documentId) {
  const getDocumentLinkFunc = getGlobalWithDtRuntime().dtRuntime?.navigation?.getDocumentLink;
  if (typeof getDocumentLinkFunc === "function") {
    return getDocumentLinkFunc(documentId);
  } else {
    console.warn("Missing getDocumentLink function from web runtime.");
    return "https://dynatrace.com/";
  }
}

// packages/platform/navigation/src/lib/open-document.ts
function openDocument(documentId) {
  const openDocumentFunc = getGlobalWithDtRuntime().dtRuntime?.navigation?.openDocument;
  if (typeof openDocumentFunc === "function") {
    openDocumentFunc(documentId);
  } else {
    console.warn("Missing openDocument function from web runtime.");
  }
}

// packages/platform/navigation/src/lib/types/builtin/dt-entity-host.ts
var getDtEntityHostDefinition = (required) => ({
  "dt.entity.host": {
    required,
    schema: {
      type: "string"
    }
  }
});

// packages/platform/navigation/src/lib/types/builtin/dt-timeframe.ts
var getDtTimeframeDefinition = (required) => ({
  "dt.timeframe": {
    required,
    schema: {
      type: "object",
      required: ["from", "to"],
      properties: {
        from: {
          type: "string"
        },
        to: {
          type: "string"
        }
      }
    }
  }
});

// packages/platform/navigation/src/lib/types/builtin/dt-query.ts
var getDtQueryDefinition = (required) => ({
  "dt.query": {
    required,
    schema: {
      type: "string"
    }
  }
});

// packages/platform/navigation/src/lib/types/intent-declaration.ts
var builtinTypeToDeclaration = {
  "dt.entity.host": getDtEntityHostDefinition,
  "dt.timeframe": getDtTimeframeDefinition,
  "dt.query": getDtQueryDefinition
};
var getIntentPropertyDeclaration = (field, required) => {
  return builtinTypeToDeclaration[field](required);
};
export {
  getAppLink,
  getDocumentLink,
  getIntent,
  getIntentLink,
  getIntentPropertyDeclaration,
  openApp,
  openDocument,
  sendIntent,
  sendIntentWithResponse,
  setPathChangeHandler
};
