// @dynatrace/strato-components/esm/charts/honeycomb/components/canvas/renderers/CanvasCircleRenderer.js
import { useCallback } from "react";
import { useCanvas } from "../../../../core/components/canvas/hooks/useCanvas.js";
import { drawLabelAndValue } from "../utils/draw-label-and-value.js";
import { fillShape } from "../utils/fill-shape.js";
import { removeTransparentBorder } from "../utils/remove-transparent-border.js";
const createCirclePath = (ctx, radius, x, y) => {
  ctx.beginPath();
  const safeRadius = Math.max(radius, 0);
  ctx.arc(x, y, safeRadius, 0, 2 * Math.PI);
  ctx.closePath();
};
const getCircleTextArea = (x, y, squareSideSize) => ({
  x,
  y,
  width: squareSideSize,
  height: squareSideSize - 8,
  leftBoundary: x - squareSideSize / 2,
  rightBoundary: x + squareSideSize / 2
});
const getCircumscribedSquare = (radius) => radius * Math.sqrt(2);
const CanvasCircleRenderer = (props) => {
  const {
    x,
    y,
    radius,
    fillColor,
    strokeColor = fillColor,
    value,
    name = "",
    showLabels,
    truncationMode
  } = props;
  const squareSideSize = getCircumscribedSquare(radius);
  const textArea = getCircleTextArea(x, y, squareSideSize);
  const draw = useCallback(
    (ctx) => {
      createCirclePath(ctx, radius, x, y);
      removeTransparentBorder(ctx, strokeColor);
      fillShape(ctx, fillColor);
      if (showLabels) {
        drawLabelAndValue(
          ctx,
          value,
          textArea,
          fillColor,
          name,
          truncationMode
        );
      }
      return false;
    },
    [
      x,
      y,
      radius,
      fillColor,
      strokeColor,
      showLabels,
      truncationMode,
      name,
      value
    ]
  );
  useCanvas("honeycomb-circle", draw);
  return null;
};
CanvasCircleRenderer["displayName"] = "CanvasCircleRenderer";
export {
  CanvasCircleRenderer
};
