// @dynatrace/strato-components/esm/charts/honeycomb/components/canvas/renderers/CanvasSquareRenderer.js
import { useCallback } from "react";
import { useCanvas } from "../../../../core/components/canvas/hooks/useCanvas.js";
import { drawLabelAndValue } from "../utils/draw-label-and-value.js";
import { fillShape } from "../utils/fill-shape.js";
import { removeTransparentBorder } from "../utils/remove-transparent-border.js";
const createRoundedSquarePath = (ctx, x, y, width, radius) => {
  ctx.moveTo(x, y);
  ctx.beginPath();
  ctx.roundRect(x, y, width, width, radius);
};
const getSquareTextArea = (x, y, width) => ({
  x: x + width / 2,
  y: y + width / 2,
  width,
  height: width,
  leftBoundary: x,
  rightBoundary: x + width
});
const CanvasSquareRenderer = (props) => {
  const {
    x,
    y,
    width,
    fillColor,
    strokeColor = fillColor,
    value,
    name = "",
    radius = 0,
    showLabels,
    truncationMode
  } = props;
  const textArea = getSquareTextArea(x, y, width);
  const draw = useCallback(
    (ctx) => {
      createRoundedSquarePath(ctx, x, y, width, radius);
      removeTransparentBorder(ctx, strokeColor);
      fillShape(ctx, fillColor);
      if (showLabels) {
        drawLabelAndValue(
          ctx,
          value,
          textArea,
          fillColor,
          name,
          truncationMode
        );
      }
      return false;
    },
    [
      x,
      y,
      width,
      fillColor,
      strokeColor,
      showLabels,
      truncationMode,
      name,
      value,
      radius
    ]
  );
  useCanvas("honeycomb-square", draw);
  return null;
};
CanvasSquareRenderer["displayName"] = "CanvasSquareRenderer";
export {
  CanvasSquareRenderer
};
