// @dynatrace/strato-components/filters/filter-field/utils/suggestions-utils/get-suggestion-category.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var get_suggestion_category_exports = {};
__export(get_suggestion_category_exports, {
  SHOW_OPERATOR_REGEX: () => SHOW_OPERATOR_REGEX,
  SPLIT_TOKEN_REGEX: () => SPLIT_TOKEN_REGEX,
  getSuggestionCategory: () => getSuggestionCategory
});
module.exports = __toCommonJS(get_suggestion_category_exports);
var import_get_text = require("../get-text.js");
var import_tree_node_guards = require("../tree-node-guards.js");
function getSuggestionCategory(args) {
  const {
    currentStatement,
    currentToken,
    currentTokenType,
    prevNode,
    tree,
    doc,
    cursorPosition
  } = args;
  let treeCursor = tree.cursorAt(cursorPosition, -1);
  if (treeCursor.node.name === "Filter") {
    treeCursor = tree.cursorAt(cursorPosition, 1);
  }
  const isFreeText = treeCursor.node.name === "FreeText";
  treeCursor.childBefore(cursorPosition);
  let showInclusionOperators = false;
  let showLogicalOperators = false;
  let suggestionType = "key";
  if (
    // Not at the first token, first token inside group and either:
    // - no statement at the current position e.g. "a = b |"
    // - started typing a key e.g. "a = b o|"
    // - just typed a logical operator, but no space yet e.g. "a = b or|"
    // -> show logical operator suggestions. (if the key that is typed does not start with o, the
    // or suggestion will be removed when filtering the suggestions)
    currentToken?.range.from !== 0 && // Case 1: no statement at the current position e.g. "a = b |"
    (currentStatement === void 0 && prevNode !== "LogicalOperator" && prevNode !== "FilterKey" && prevNode !== "ComparisonOperator" && prevNode !== "OpeningBracket" || // Case 2: started typing a key e.g. "a = b o|"
    !!currentStatement && currentTokenType === "key" && prevNode !== "LogicalOperator" && treeCursor?.node?.parent?.node?.prevSibling?.name !== "OpeningBracket" || // Case 3: just typed a logical operator, but no space yet e.g. "a = b or|"
    prevNode === "LogicalOperator" && !!currentToken?.textValue && treeCursor?.node?.prevSibling?.name !== "OpeningBracket")
  ) {
    showLogicalOperators = true;
  } else if (currentTokenType === "value" && !Array.isArray(currentToken.value) || currentToken.type === "String" && !currentToken.value && currentStatement && !currentStatement.value && prevNode !== "FilterKey") {
    suggestionType = "value";
  } else if (currentToken.type === "Key" && (currentStatement?.comparisonOperator !== void 0 && currentStatement.comparisonOperator.type !== "Error" || currentStatement?.value !== void 0)) {
    suggestionType = "key";
  } else if (prevNode === "FilterKey" && (currentToken.type === "String" || currentToken.type === "Error") || currentToken.type === "ComparisonOperator" || currentToken.type === "InclusionOperator" || currentToken.type === "ExistsOperator" || prevNode === "InvalidComparisonOperator" && cursorPosition <= (currentStatement?.comparisonOperator?.range.to ?? 0) || currentToken.type === "Error" && currentTokenType === "comparisonOperator") {
    suggestionType = "comparisonOperator";
  } else if (
    // List suggestions need to be handled separately since we don't want to replace the whole list
    // token, but only the list values when applying suggestions. Only the list value at the cursor
    // position should be returned as `currentToken` that can be used for filtering suggestions.
    Array.isArray(currentToken.value) && (0, import_tree_node_guards.isFilterFieldListNode)(currentToken)
  ) {
    suggestionType = "value";
  }
  if (prevNode === "ComparisonOperator" && cursorPosition > treeCursor.node.to && (currentStatement?.comparisonOperator?.range.to ?? 0) < (currentStatement?.value?.range.from ?? 0) && currentTokenType !== "comparisonOperator" || prevNode === "ComparisonOperator" && !currentStatement || prevNode === "FilterValue" || prevNode === "\u26A0" && currentTokenType === "value") {
    suggestionType = "value";
  }
  if (prevNode === "InclusionOperator" && cursorPosition > treeCursor.node.to && treeCursor.nextSibling() && treeCursor.node.name === "FilterIncludesValue") {
    const { from } = treeCursor.node;
    const characterBeforeCursor = (0, import_get_text.getTextFromRange)(
      {
        from: from - 1,
        to: from
      },
      doc
    );
    if (characterBeforeCursor === " ") {
      showInclusionOperators = true;
    }
    suggestionType = "none";
  }
  if (prevNode === "OpeningBracket" && treeCursor.nextSibling() && treeCursor.node.name === "\u26A0") {
    showInclusionOperators = true;
  }
  if (prevNode === "ExistsOperator" && currentTokenType === "comparisonOperator") {
    if (cursorPosition < currentToken.range.to) {
      suggestionType = "comparisonOperator";
    } else {
      suggestionType = "value";
    }
  }
  let overlappingSuggestions = void 0;
  if (isFreeText && suggestionType === "key") {
    overlappingSuggestions = getOverlappingSuggestionTypes(
      currentToken,
      cursorPosition
    );
  }
  return {
    suggestionType,
    showLogicalOperators,
    showInclusionOperators,
    overlappingSuggestions
  };
}
const SPLIT_TOKEN_REGEX = /(?<!\\)(?:\\\\)*  */g;
const SHOW_OPERATOR_REGEX = /(?<!\\)(?:\\\\)*((?<rightAssociative>[ !])| (?<keyword>i|n|in|no|not|not i|not in) ?)$/d;
function getOverlappingSuggestionTypes(currentToken, cursorPosition) {
  const { range, textValue } = currentToken;
  const neutralizedCursorPosition = cursorPosition - range.from;
  let subTokenSeparator = null;
  let nextMatch = null;
  const matchSubToken = new RegExp(SPLIT_TOKEN_REGEX, "g");
  while ((nextMatch = matchSubToken.exec(textValue)) !== null) {
    subTokenSeparator = nextMatch;
  }
  const operatorSeparators = new RegExp(SHOW_OPERATOR_REGEX).exec(textValue);
  if (!subTokenSeparator && !operatorSeparators) {
    return;
  }
  if (operatorSeparators) {
    const offsetSpace = textValue.slice(-1) === " " ? 1 : 0;
    const {
      keyword,
      rightAssociative
    } = operatorSeparators.groups ?? {};
    const trimmedTokenValue = (rightAssociative ?? "").trim();
    const {
      keyword: keywordIndices,
      rightAssociative: rightAssociativeIndices
    } = operatorSeparators.indices?.groups ?? {};
    if (neutralizedCursorPosition >= keywordIndices?.[0]) {
      return {
        type: "comparisonOperator",
        subToken: {
          type: "String",
          value: keyword ?? trimmedTokenValue,
          textValue: keyword ?? trimmedTokenValue,
          range: {
            from: range.from + keywordIndices[0],
            to: range.from + keywordIndices[1] + offsetSpace
          }
        }
      };
    }
    if (neutralizedCursorPosition > rightAssociativeIndices?.[0]) {
      return {
        type: "comparisonOperator",
        subToken: {
          type: "String",
          value: keyword ?? trimmedTokenValue,
          textValue: keyword ?? trimmedTokenValue,
          range: {
            from: range.from + rightAssociativeIndices[0] + offsetSpace,
            to: range.from + rightAssociativeIndices[1]
          }
        }
      };
    }
  }
}
