// @dynatrace/strato-components/filters/filter-field/utils/suggestions-utils/get-current-token.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var get_current_token_exports = {};
__export(get_current_token_exports, {
  getCurrentToken: () => getCurrentToken
});
module.exports = __toCommonJS(get_current_token_exports);
var import_tree_node_guards = require("../tree-node-guards.js");
var import_suggestions_callback = require("./suggestions-callback.js");
function getCurrentToken({
  tree,
  convertedTree,
  cursorPosition,
  doc
}) {
  const fallbackToken = {
    type: "String",
    textValue: "",
    value: "",
    range: {
      from: cursorPosition,
      to: cursorPosition
    }
  };
  let currentToken = fallbackToken;
  let currentTokenType = void 0;
  const currentStatement = (0, import_suggestions_callback.getFilterStatement)(
    convertedTree,
    cursorPosition,
    doc
  );
  const treeCursor = tree.cursorAt(cursorPosition, -1);
  if (!currentStatement) {
    if (treeCursor.node.name === "LogicalOperator") {
      currentToken = convertedTree?.children.find(
        (token) => (0, import_tree_node_guards.isFilterFieldValueNode)(token) && token.range.from <= cursorPosition && token.range.to >= cursorPosition
      ) ?? fallbackToken;
    }
  } else {
    for (const [key, value] of Object.entries(currentStatement)) {
      if ((0, import_tree_node_guards.isFilterFieldValueNode)(value) && value.range.from <= cursorPosition && value.range.to >= cursorPosition) {
        currentToken = value;
        currentTokenType = key;
        break;
      }
    }
  }
  if (Array.isArray(currentToken.value) && (0, import_tree_node_guards.isFilterFieldListNode)(currentToken)) {
    currentToken = currentToken.value?.find(
      (token) => (0, import_tree_node_guards.isFilterFieldLeafNode)(token) && token.range.from <= cursorPosition && token.range.to >= cursorPosition
    ) ?? currentToken;
    if (currentToken.type === "List" || currentToken.type === "Error") {
      let nodeCursor = tree.cursorAt(cursorPosition, 1);
      if (nodeCursor.node.name === "Filter") {
        nodeCursor = tree.cursorAt(cursorPosition, -1);
      }
      const { node } = nodeCursor.node;
      const tokensInsideList = [
        "List",
        "Identifier",
        "String",
        "Boolean",
        "Number",
        "Duration"
      ];
      if (tokensInsideList.includes(node.name) || node.name === "OpeningBracket" && cursorPosition > node.from || node.name === "ClosingBracket" && cursorPosition < node.to) {
        currentToken = fallbackToken;
      }
    }
  }
  return { currentToken, currentTokenType };
}
