// @dynatrace/strato-components/filters/filter-field/utils/suggestions-utils/key-suggestions.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var key_suggestions_exports = {};
__export(key_suggestions_exports, {
  JSONPATH_PATTERN_PARTIAL: () => JSONPATH_PATTERN_PARTIAL,
  getFallbackKeySuggestions: () => getFallbackKeySuggestions,
  getFallbackKeys: () => getFallbackKeys,
  getFilterStatementKeysMap: () => getFilterStatementKeysMap,
  getFilterStatementSuggestions: () => getFilterStatementSuggestions,
  getKeySuggestions: () => getKeySuggestions,
  getSearchSuggestions: () => getSearchSuggestions
});
module.exports = __toCommonJS(key_suggestions_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_shared_suggestions = require("./shared-suggestions.js");
var import_suggestion_type = require("./suggestion-type.js");
var import_messages = require("../../messages.js");
var import_normalize_validator_map = require("../validation-utils/normalize-validator-map.js");
var import_type_guards = require("../validation-utils/type-guards.js");
var import_validation_helper_functions = require("../validation-utils/validation-helper-functions.js");
const JSONPATH_PATTERN_PARTIAL = /^(?:"(?:[^"\\]|\\.)*"|[^\s"$]+)\$(?:(?:\.(?:[a-zA-Z_][\w]*|"(?:[^"\\]|\\.)*")|\[(?:\d+|"(?:[^"\\]|\\.)*")\])*)/;
const SEARCH_OPERATOR_VALUES = ["", "*", "* ", "*~", "* ~"];
const isSearchOperatorValue = (value) => SEARCH_OPERATOR_VALUES.includes(value);
function getKeySuggestions(internalValidatorMap, suggestionTypes = [], customTypes, doc, currentToken) {
  const validatorMapIsExhaustive = internalValidatorMap?.exhaustive;
  let textValue = currentToken?.textValue ?? "";
  const hasTrailingSpace = textValue.endsWith(" ");
  const textValueNeedsEscaping = textValue.toLocaleLowerCase() === "or" || textValue.toLocaleLowerCase() === "and";
  if (textValue && textValueNeedsEscaping && doc && currentToken) {
    textValue = doc.slice(currentToken.range.from, currentToken.range.to).toString();
  }
  if (!internalValidatorMap) {
    if (textValue && !validatorMapIsExhaustive && !hasTrailingSpace && !isSearchOperatorValue(textValue)) {
      return [
        {
          id: import_normalize_validator_map.KEY_GROUP_ID,
          suggestions: [
            getDefaultKeySuggestion(textValue, textValueNeedsEscaping)
          ]
        }
      ];
    } else if (!textValue && !validatorMapIsExhaustive) {
      return [
        { id: import_normalize_validator_map.KEY_GROUP_ID, suggestions: [(0, import_shared_suggestions.getStartTypingSuggestion)("key")] }
      ];
    }
    return [];
  }
  const { byKey, keyToGroup } = internalValidatorMap.normalizedKeys;
  const groupEntryById = /* @__PURE__ */ new Map();
  const result = [];
  let suggestionMatchesTextValue = false;
  for (const [key, config] of byKey) {
    if (key === textValue) {
      suggestionMatchesTextValue = true;
    }
    let suggestionTypeResult;
    if (config.valueType) {
      suggestionTypeResult = (0, import_suggestion_type.getValidValueTypes)({ type: config.valueType });
    } else if (config.valuePredicate) {
      suggestionTypeResult = (0, import_suggestion_type.getValidValueTypes)(config.valuePredicate);
    } else {
      suggestionTypeResult = {
        suggestionType: "Any",
        hasLiteralValues: false,
        types: ["Any"]
      };
    }
    const { suggestionType, hasLiteralValues } = suggestionTypeResult;
    const displayValue = config.displayValue;
    const details = config.details;
    let keySuggestion;
    if (suggestionType === "JSONPath") {
      keySuggestion = getJSONPathKeySuggestion(key, suggestionType);
    } else if (Array.isArray(suggestionType)) {
      keySuggestion = suggestionType.map((type) => {
        if (type === "JSONPath") {
          return getJSONPathKeySuggestion(key, type);
        }
        return {
          value: key,
          icon: getIconForValueType(type),
          suggestionType: "validatorMap",
          displayValue,
          details
        };
      });
    } else if (hasLiteralValues && Array.isArray(config.valuePredicate)) {
      keySuggestion = {
        value: key,
        icon: getIconForValueType("Any"),
        suggestionType: "validatorMap",
        displayValue,
        details
      };
    } else {
      keySuggestion = {
        value: key,
        icon: getIconForValueType(suggestionType, customTypes),
        suggestionType: "validatorMap",
        displayValue,
        details
      };
    }
    const keySuggestions = Array.isArray(keySuggestion) ? keySuggestion : [keySuggestion];
    const group = keyToGroup.get(key);
    if (group === void 0) {
      continue;
    }
    let groupEntry = groupEntryById.get(group.id);
    if (!groupEntry) {
      groupEntry = {
        id: group.id,
        label: group.label,
        "aria-label": group["aria-label"],
        suggestions: []
      };
      groupEntryById.set(group.id, groupEntry);
      result.push(groupEntry);
    }
    groupEntry.suggestions.push(...keySuggestions);
  }
  const isJSONPathContinuation = currentToken?.type === "JSONPath" || currentToken?.type === "Error" && JSONPATH_PATTERN_PARTIAL.test(currentToken?.textValue ?? "");
  if (textValue && !validatorMapIsExhaustive && !suggestionMatchesTextValue && !isSearchOperatorValue(textValue) && (!hasTrailingSpace || suggestionTypes.includes("pastedContent")) && !isJSONPathContinuation) {
    const trailingSuggestion = getDefaultKeySuggestion(
      textValue,
      textValueNeedsEscaping
    );
    const lastGroup = result.at(-1);
    if (lastGroup?.id.startsWith(`${import_normalize_validator_map.KEY_GROUP_ID}-implicit-`)) {
      lastGroup.suggestions.push(trailingSuggestion);
    } else {
      result.push({
        id: `${import_normalize_validator_map.KEY_GROUP_ID}-implicit-${result.length}`,
        suggestions: [trailingSuggestion]
      });
    }
  }
  return result;
}
function getSearchSuggestions(context, currentStatement, currentToken, suggestionTypes, isInsideAGroup) {
  if (isInsideAGroup) {
    return [];
  }
  const textValue = currentToken.textValue;
  const { value } = currentToken;
  if (currentStatement?.comparisonOperator !== void 0 && currentStatement.comparisonOperator.type !== "SearchOperator" || currentStatement?.comparisonOperator?.type === "SearchOperator" && value === "") {
    return [];
  }
  const emptyInput = context.state.doc.toString() === "";
  if (emptyInput || isSearchOperatorValue(textValue)) {
    return [(0, import_shared_suggestions.getSearchSuggestion)()];
  } else if ((suggestionTypes.includes("key") || suggestionTypes.includes("value") || suggestionTypes.includes("pastedContent")) && !(currentToken.type === "JSONPath" || currentToken.type === "Error" && JSONPATH_PATTERN_PARTIAL.exec(currentToken.textValue) && !currentToken.textValue.endsWith("$"))) {
    return [(0, import_shared_suggestions.getSearchSuggestion)(textValue)];
  }
  return [];
}
function getFallbackKeySuggestions(currentToken, suggestionTypes, fallbackKeys, filterStatementSuggestions) {
  if (!fallbackKeys || fallbackKeys.length === 0) {
    return [];
  }
  const { value } = currentToken;
  const suggestions = [];
  if (suggestionTypes.includes("key") && value !== "") {
    const textValue = currentToken.textValue;
    for (const key of fallbackKeys) {
      if (Array.isArray(key.comparisonOperators)) {
        for (const operator of key.comparisonOperators) {
          const fallbackKeySuggestion = (0, import_shared_suggestions.getFallbackKeySuggestion)(
            key.value,
            operator,
            textValue
          );
          if (!(operator === "equals" && filterStatementSuggestions?.some(
            (statementSuggestion) => statementSuggestion.value === fallbackKeySuggestion.value
          ))) {
            suggestions.push(fallbackKeySuggestion);
          }
        }
      } else {
        suggestions.push(
          (0, import_shared_suggestions.getFallbackKeySuggestion)(key.value, "matches-phrase", textValue)
        );
      }
    }
  }
  return suggestions;
}
function getFilterStatementSuggestions(currentToken, suggestionTypes, filterStatementKeysMap) {
  const { textValue } = currentToken;
  if (!filterStatementKeysMap || !suggestionTypes.includes("key") || textValue === "") {
    return [];
  }
  const suggestions = [];
  if (filterStatementKeysMap.values.size > 0) {
    const lowerTextValue = textValue.toLowerCase();
    for (const [value, keys] of filterStatementKeysMap.values.entries()) {
      if (value.toLowerCase().includes(lowerTextValue)) {
        for (const key of keys) {
          suggestions.push((0, import_shared_suggestions.getStatementSuggestionForMatchingValue)(key, value));
        }
      }
    }
  }
  if (filterStatementKeysMap.functions.size > 0) {
    for (const [key, func] of filterStatementKeysMap.functions.entries()) {
      if (func(currentToken)) {
        suggestions.push(
          (0, import_shared_suggestions.getStatementSuggestionForMatchingValue)(key, currentToken.textValue)
        );
      }
    }
  }
  return suggestions;
}
function getFilterStatementKeysMap(validatorMap) {
  if (!validatorMap || Array.isArray(validatorMap.keyPredicates) || !validatorMap.keyPredicates) {
    return void 0;
  }
  const valuesMap = /* @__PURE__ */ new Map();
  const functionsMap = /* @__PURE__ */ new Map();
  for (const [key, keyPredicate] of Object.entries(
    validatorMap.keyPredicates
  )) {
    if (!(0, import_type_guards.isFilterFieldValidatorMapTypeSpecification)(keyPredicate)) {
      if (typeof keyPredicate.suggestStatementOnValueMatch === "function") {
        functionsMap.set(key, keyPredicate.suggestStatementOnValueMatch);
      } else if (keyPredicate.suggestStatementOnValueMatch && keyPredicate.valuePredicate) {
        const valuePredicate = keyPredicate.valuePredicate;
        if (Array.isArray(valuePredicate)) {
          for (const value of valuePredicate) {
            if (!(0, import_type_guards.isFilterFieldValueSuggestionGroupConfig)(value)) {
              addKeyValuePairToMap(valuesMap, key, value);
            }
          }
        } else if (valuePredicate && typeof valuePredicate !== "function" && !(0, import_type_guards.isFilterFieldValueSuggestionGroupConfig)(valuePredicate)) {
          addKeyValuePairToMap(valuesMap, key, valuePredicate);
        }
      }
    }
  }
  return valuesMap.size > 0 || functionsMap.size > 0 ? { values: valuesMap, functions: functionsMap } : void 0;
}
function addKeyValuePairToMap(valuesMap, key, value) {
  if (!(0, import_type_guards.isFilterFieldValidatorMapTypeSpecification)(value)) {
    const stringValue = (0, import_validation_helper_functions.getValuePredicateTextValue)(value);
    const existingKeys = valuesMap.get(stringValue) ?? [];
    if (!existingKeys.includes(key)) {
      valuesMap.set(stringValue, [...existingKeys, key]);
    }
  }
}
function getFallbackKeys(validatorMap) {
  if (!validatorMap || Array.isArray(validatorMap.keyPredicates)) {
    return void 0;
  }
  const keys = [];
  for (const key in validatorMap.keyPredicates) {
    const keyPredicate = validatorMap.keyPredicates[key];
    if (!(0, import_type_guards.isFilterFieldValidatorMapTypeSpecification)(keyPredicate) && keyPredicate.fallbackKey) {
      keys.push({
        value: key,
        comparisonOperators: Array.isArray(keyPredicate.fallbackKey) ? keyPredicate.fallbackKey : void 0
      });
    }
  }
  return keys.length > 0 ? keys : void 0;
}
function getIconForCustomType(type, customTypes) {
  for (const [typeName, customType] of Object.entries(customTypes)) {
    if (typeName === type && typeof customType === "object" && customType.icon !== void 0) {
      return customType.icon;
    }
  }
  return null;
}
function getIconForValueType(type, customTypes) {
  switch (type) {
    case "Number": {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.HashtagIcon, {});
    }
    case "Boolean": {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.BooleanIcon, {});
    }
    case "Duration": {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.StopwatchIcon, {});
    }
    case "JSONPath": {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.JsonIcon, {});
    }
    case "IPAddress": {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.IPaddressIcon, {});
    }
    case "UID": {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.UIDIcon, {});
    }
    case "SmartscapeId": {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.UIDIcon, {});
    }
    case "Timestamp": {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ClockIcon, {});
    }
    case "String":
    case "Any":
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.AbcIcon, {});
  }
  if (customTypes) {
    return getIconForCustomType(type, customTypes);
  }
}
function getDefaultKeySuggestion(value, valueNeedsEscaping = false) {
  const match = /^"(.+)(?<!\\)"$/.exec(value);
  const textValue = match ? match[1] : value;
  return {
    trailing: " ",
    suggestionType: "hint-useAsKey",
    value: valueNeedsEscaping ? `"${value}"` : value,
    id: "strato-filterfield-suggestion-use-as-key",
    displayValue: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_react_intl.FormattedMessage,
      {
        ...import_messages.filterFieldMessages.suggestions.useAsKey,
        values: {
          value: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
            "\xA0",
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)("mark", { children: textValue }),
            "\xA0"
          ] })
        }
      }
    )
  };
}
function getJSONPathKeySuggestion(key, valueType = "Any") {
  return {
    value: `${key}$.`,
    icon: getIconForValueType(valueType),
    trailing: "",
    suggestionType: "JSONPath",
    cursorBeforeTrailing: true,
    details: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_react_intl.FormattedMessage,
      {
        defaultMessage: "Filter JSON attributes",
        description: "Details shown for JSONPath key suggestions.",
        id: "/SFvkA0WT3AVdLhH"
      }
    ),
    insertionStrategy: "replace-statement"
  };
}
