// @dynatrace/strato-components/filters/filter-field/utils/suggestions-utils/operator-suggestions.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var operator_suggestions_exports = {};
__export(operator_suggestions_exports, {
  ValuePlaceholder: () => ValuePlaceholder,
  defaultLogicalOperatorSuggestions: () => defaultLogicalOperatorSuggestions,
  getComparisonOperatorSuggestions: () => getComparisonOperatorSuggestions,
  operatorSuggestionsWithDetails: () => operatorSuggestionsWithDetails
});
module.exports = __toCommonJS(operator_suggestions_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_intl = require("react-intl");
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
var import_typography = __toESM(require("@dynatrace/strato-design-tokens/typography"));
var import_messages = require("../../messages.js");
var import_validation_helper_functions = require("../validation-utils/validation-helper-functions.js");
function getComparisonOperatorSuggestions(key, internalValidatorMap, parserConfig) {
  const operators = (0, import_validation_helper_functions.getValidOperatorsShort)(
    key,
    internalValidatorMap,
    parserConfig
  );
  return operators.map((operator) => {
    const operatorWithDetails = operatorSuggestionsWithDetails[operator];
    const baseProps = {
      suggestionType: "validatorMap"
    };
    return operatorWithDetails ? { ...operatorWithDetails, ...baseProps } : { value: operator, ...baseProps };
  });
}
const defaultLogicalOperatorSuggestions = [
  { value: "OR" }
];
const ValuePlaceholder = () => {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "span",
    {
      style: {
        color: import_colors.default.Text.Neutral.Subdued,
        fontWeight: import_typography.default.Text.Base.Default.Weight
      },
      children: "value"
    }
  );
};
const operatorSuggestionsWithDetails = {
  "=": {
    value: "=",
    details: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_intl.FormattedMessage, { ...import_messages.filterFieldMessages.comparisonOperators.equals }),
    escapeSuggestion: false
  },
  "!=": {
    value: "!=",
    details: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_react_intl.FormattedMessage,
      {
        ...import_messages.filterFieldMessages.comparisonOperators.notEquals
      }
    ),
    escapeSuggestion: false
  },
  "<": {
    value: "<",
    details: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_intl.FormattedMessage, { ...import_messages.filterFieldMessages.comparisonOperators.lessThan }),
    escapeSuggestion: false
  },
  "<=": {
    value: "<=",
    details: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_react_intl.FormattedMessage,
      {
        ...import_messages.filterFieldMessages.comparisonOperators.lessThanEqual
      }
    ),
    escapeSuggestion: false
  },
  ">": {
    value: ">",
    details: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_react_intl.FormattedMessage,
      {
        ...import_messages.filterFieldMessages.comparisonOperators.greaterThan
      }
    ),
    escapeSuggestion: false
  },
  ">=": {
    value: ">=",
    details: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_react_intl.FormattedMessage,
      {
        ...import_messages.filterFieldMessages.comparisonOperators.greaterThanEqual
      }
    ),
    escapeSuggestion: false
  },
  in: {
    value: "in (",
    displayValue: "in ()",
    details: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_intl.FormattedMessage, { ...import_messages.filterFieldMessages.comparisonOperators.in }),
    trailing: ") ",
    cursorBeforeTrailing: true,
    escapeSuggestion: false
  },
  "not in": {
    value: "not in (",
    displayValue: "not in ()",
    details: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_intl.FormattedMessage, { ...import_messages.filterFieldMessages.comparisonOperators.notIn }),
    trailing: ") ",
    cursorBeforeTrailing: true,
    escapeSuggestion: false
  },
  "starts-with": {
    value: '= "',
    displayValue: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { children: [
      "= ",
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(ValuePlaceholder, {}),
      "*"
    ] }),
    details: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_react_intl.FormattedMessage,
      {
        ...import_messages.filterFieldMessages.comparisonOperators.startsWith
      }
    ),
    trailing: 'value"*',
    cursorBeforeTrailing: true,
    escapeSuggestion: false
  },
  contains: {
    value: '= *"',
    displayValue: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { children: [
      "= *",
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(ValuePlaceholder, {}),
      "*"
    ] }),
    details: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_intl.FormattedMessage, { ...import_messages.filterFieldMessages.comparisonOperators.contains }),
    trailing: 'value"*',
    cursorBeforeTrailing: true,
    escapeSuggestion: false
  },
  "ends-with": {
    value: '= *"',
    displayValue: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { children: [
      "= *",
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(ValuePlaceholder, {})
    ] }),
    details: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_intl.FormattedMessage, { ...import_messages.filterFieldMessages.comparisonOperators.endsWith }),
    trailing: 'value"',
    cursorBeforeTrailing: true,
    escapeSuggestion: false
  },
  "not-starts-with": {
    value: '!= "',
    displayValue: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { children: [
      "!= ",
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(ValuePlaceholder, {}),
      "*"
    ] }),
    details: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_react_intl.FormattedMessage,
      {
        ...import_messages.filterFieldMessages.comparisonOperators.notStartsWith
      }
    ),
    trailing: 'value"*',
    cursorBeforeTrailing: true,
    escapeSuggestion: false
  },
  "not-contains": {
    value: '!= *"',
    displayValue: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { children: [
      "!= *",
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(ValuePlaceholder, {}),
      "*"
    ] }),
    details: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_react_intl.FormattedMessage,
      {
        ...import_messages.filterFieldMessages.comparisonOperators.notContains
      }
    ),
    trailing: 'value"*',
    cursorBeforeTrailing: true,
    escapeSuggestion: false
  },
  "not-ends-with": {
    value: '!= *"',
    displayValue: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { children: [
      "!= *",
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(ValuePlaceholder, {})
    ] }),
    details: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_react_intl.FormattedMessage,
      {
        ...import_messages.filterFieldMessages.comparisonOperators.notEndsWith
      }
    ),
    trailing: 'value"',
    cursorBeforeTrailing: true,
    escapeSuggestion: false
  },
  "= *": {
    value: "= *",
    details: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_intl.FormattedMessage, { ...import_messages.filterFieldMessages.comparisonOperators.exists })
  },
  "!= *": {
    value: "!= *",
    details: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_react_intl.FormattedMessage,
      {
        ...import_messages.filterFieldMessages.comparisonOperators.notExists
      }
    ),
    escapeSuggestion: false
  },
  "~": {
    value: "~",
    details: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_react_intl.FormattedMessage,
      {
        ...import_messages.filterFieldMessages.comparisonOperators.matchesPhrase
      }
    ),
    escapeSuggestion: false
  },
  "!~": {
    value: "!~",
    details: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_react_intl.FormattedMessage,
      {
        ...import_messages.filterFieldMessages.comparisonOperators.notMatchesPhrase
      }
    ),
    escapeSuggestion: false
  }
};
