// @dynatrace/strato-components/filters/filter-field/utils/suggestions-utils/recent-and-pinned-suggestions.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var recent_and_pinned_suggestions_exports = {};
__export(recent_and_pinned_suggestions_exports, {
  fuzzySearchFilterSuggestion: () => fuzzySearchFilterSuggestion,
  getRecentAndPinnedSuggestions: () => getRecentAndPinnedSuggestions
});
module.exports = __toCommonJS(recent_and_pinned_suggestions_exports);
var import_editor_config = require("../../editor-state-configs/editor-config.js");
function mergeRanges(ranges) {
  if (ranges.length === 0) {
    return [];
  }
  ranges.sort((a, b) => a[0] - b[0]);
  const merged = [];
  let [start, end] = ranges[0];
  for (const [nextStart, nextEnd] of ranges) {
    if (nextStart <= end + 1) {
      end = Math.max(end, nextEnd);
    } else {
      merged.push([start, end]);
      start = nextStart;
      end = nextEnd;
    }
  }
  merged.push([start, end]);
  return merged;
}
function createNeedleMatchPattern(needle) {
  const savePattern = needle.toLowerCase().replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  const operators = [
    { char: /\s*!=\s*\\\*/g, placeholder: "\u03F40", pattern: "\\s*!=\\s*\\*" },
    { char: /\s*=\s*\\\*/g, placeholder: "\u03F41", pattern: "\\s*=\\s*\\*" },
    { char: /\s*<=\s*/, placeholder: "\u03F42", pattern: "\\s*<=\\s*" },
    { char: /\s*>=\s*/, placeholder: "\u03F43", pattern: "\\s*>=\\s*" },
    { char: /\s*!=\s*/, placeholder: "\u03F44", pattern: "\\s*!=\\s*" },
    { char: /\s*!~\s*/, placeholder: "\u03F45", pattern: "\\s*!~\\s*" },
    { char: /\s*=\s*/, placeholder: "\u03F46", pattern: "\\s*=\\s*" },
    { char: /\s*<\s*/, placeholder: "\u03F47", pattern: "\\s*<\\s*" },
    { char: /\s*>\s*/, placeholder: "\u03F48", pattern: "\\s*>\\s*" },
    { char: /\s*~\s*/, placeholder: "\u03F49", pattern: "\\s*~\\s*" }
  ];
  const listOperator = [
    /*
     * Breakdown:
     * - (not )?         — optionally matches "not " before "in"
     * - in              — matches "in"
     * - \s*             — matches zero or more whitespace characters
     * - (?=\s*\\\()     — positive lookahead for an escaped "("
     * Purpose: Matches "in" or "not in", only if followed by an opening parenthesis.
     * https://regex101.com/r/qW9mOD/1
     */
    {
      char: /(\s*(not\s+)?in\s*)(?=\s*\\\()/g,
      placeholder: "\u03A90",
      pattern: "\\s*(not\\s+)?in\\s*"
    },
    /*
     * Breakdown:
     * - (?<=Ω0)         — positive lookbehind for the Ω0 placeholder ("in" or "not in")
     * - \\*\(           — an escaped "("
     * - \s*             — optional whitespace characters
     * Purpose: Matches an escaped opening parenthesis and optional whitespace after "in"/"not in".
     * https://regex101.com/r/ILgKSm/1
     */
    {
      char: /(?<=Ω0)\\*\(\s*/g,
      placeholder: "\u03A91",
      pattern: "\\(\\s*"
    },
    /*
     * Breakdown:
     * - (?<=Ω0[^)]*)    — positive lookbehind for Ω0 and any number of non-")" characters
     * - \s*             — optional whitespace characters
     * - \\*\)           — an escaped ")"
     * Purpose: Matches optional spaces and an escaped closing parenthesis at the end of the IN list.
     * https://regex101.com/r/pQTBe7/1
     */
    {
      char: /(?<=Ω0[^)]*)\s*\\*\)/g,
      placeholder: "\u03A92",
      pattern: "\\s*\\)*"
    },
    /*
     * Breakdown:
     * - \s*             — optional whitespace characters before the comma
     * - ,               — matches a literal comma
     * - \s*             — optional whitespace characters after the comma
     * - (?=.*Ω2)        — positive lookahead: ensures that closing parenthesis (Ω2) exists somewhere ahead in the string
     * Purpose: Matches commas with optional whitespace inside the IN list.
     * https://regex101.com/r/Yy86Vd/1
     */
    {
      char: /\s*,\s*(?=.*Ω2)/g,
      placeholder: "\u03A93",
      pattern: "\\s*,\\s*"
    }
  ];
  const implicitLogicalAnd = [
    {
      // Regex for the logical AND matching can be lowercase as it only applies to an already lowercased string.
      char: /\sand\s/,
      placeholder: "\u03A60",
      /*
       * Breakdown:
       * - (?:...)   — non-capturing group with alternation
       * - |\s+      — OR: just one or more whitespace characters
       * - \s*AND\s+ — optional leading whitespace, then "AND", then required trailing whitespace
       *
       * This avoids nested quantifiers and handles whitespace before AND.
       * This pattern will match:
       * - " " (just whitespace)
       * - " AND " (whitespace before and after)
       * - "  AND  " (multiple spaces before and after)
       *
       * It will NOT match:
       * - "OR", "XOR", or any other logical operator
       *
       * Example matches:
       * - "foo = bar AND baz = qux"
       * - "foo = bar baz = qux"
       * - "foo = bar  AND  baz = qux"
       */
      pattern: "(?:\\s*AND\\s+|\\s+)"
    },
    { char: /\s+/g, placeholder: "\u03A61", pattern: "(?:\\s*AND\\s+|\\s+)" }
  ];
  const explicitGroup = [
    { char: /\\\(/g, placeholder: "\u03A64", pattern: "" },
    { char: /\\\)/g, placeholder: "\u03A65", pattern: "" },
    { char: /"/g, placeholder: "\u03A66", pattern: "" }
  ];
  let matchingPatterns = operators.concat(listOperator);
  matchingPatterns = matchingPatterns.concat(implicitLogicalAnd);
  matchingPatterns = matchingPatterns.concat(explicitGroup);
  let operatorOptionalWhitespacePattern = savePattern;
  for (const operator of matchingPatterns) {
    operatorOptionalWhitespacePattern = operatorOptionalWhitespacePattern.replace(
      new RegExp(operator.char, "g"),
      operator.placeholder
    );
  }
  for (const operator of matchingPatterns) {
    operatorOptionalWhitespacePattern = operatorOptionalWhitespacePattern.replace(
      new RegExp(operator.placeholder, "g"),
      operator.pattern
    );
  }
  return new RegExp(operatorOptionalWhitespacePattern, "gi");
}
function fuzzySearchFilterSuggestion(haystack, needle) {
  const result = [];
  const lowercaseNeedle = needle.toLocaleLowerCase();
  const needleMatchPattern = createNeedleMatchPattern(lowercaseNeedle);
  for (const haystackItem of haystack) {
    const { value } = haystackItem;
    const lowercaseHaystackValue = value?.toLocaleLowerCase().replace(/(?<!\s(?:not\s)?in\s*)[()"]/g, "");
    if (needleMatchPattern.source === "(?:)") {
      result.push({ suggestion: haystackItem, position: 0, ranges: [] });
      continue;
    }
    let match = null;
    if (lowercaseHaystackValue) {
      let fuzzyMatch;
      while ((fuzzyMatch = needleMatchPattern.exec(lowercaseHaystackValue)) !== null) {
        match ??= {
          suggestion: haystackItem,
          position: fuzzyMatch.index,
          ranges: []
        };
        match.ranges.push([
          fuzzyMatch.index,
          fuzzyMatch.index + fuzzyMatch[0].length
        ]);
      }
    }
    if (match !== null) {
      match.ranges = mergeRanges(match.ranges);
      result.push(match);
    }
  }
  return result;
}
function getRecentAndPinnedSuggestions(context) {
  const recentSuggestionValues = [];
  const pinnedSuggestionValues = [];
  const recentFilters = context.view?.state.field(import_editor_config.usedRecentFilters);
  const currentFilterFieldInput = context.state.doc.toString().trim();
  if (recentFilters && Array.isArray(recentFilters)) {
    let filteredRecent = fuzzySearchFilterSuggestion(
      recentFilters,
      currentFilterFieldInput
    );
    if (currentFilterFieldInput.length > 0) {
      filteredRecent = filteredRecent.slice(0, 2);
    } else {
      filteredRecent = filteredRecent.slice(0, 10);
    }
    recentSuggestionValues.push(...filteredRecent);
  }
  const pinnedFilters = context.view?.state.field(import_editor_config.usedPinnedFilters);
  if (pinnedFilters && Array.isArray(pinnedFilters)) {
    let filteredPinned = fuzzySearchFilterSuggestion(
      pinnedFilters,
      currentFilterFieldInput
    );
    if (currentFilterFieldInput.length > 0) {
      filteredPinned = filteredPinned.sort((filterA, filterB) => filterA.position - filterB.position);
    }
    pinnedSuggestionValues.push(...filteredPinned);
  }
  return { recentSuggestionValues, pinnedSuggestionValues };
}
