// @dynatrace/strato-components/filters/filter-field/utils/suggestions-utils/shared-suggestions.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var shared_suggestions_exports = {};
__export(shared_suggestions_exports, {
  getFallbackKeySuggestion: () => getFallbackKeySuggestion,
  getJSONPathHint: () => getJSONPathHint,
  getNoMatchingSuggestion: () => getNoMatchingSuggestion,
  getPastingSuggestion: () => getPastingSuggestion,
  getSearchSuggestion: () => getSearchSuggestion,
  getStartTypingSuggestion: () => getStartTypingSuggestion,
  getStatementSuggestionForMatchingValue: () => getStatementSuggestionForMatchingValue
});
module.exports = __toCommonJS(shared_suggestions_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var styles = __toESM(require("../../FilterField.sty.js"));
var import_messages = require("../../messages.js");
var suggestionStyles = __toESM(require("../../suggestions/Suggestion.sty.js"));
var import_escape_suggestion = require("../escape-suggestion.js");
var import_operator_specification = require("../validation-utils/operator-specification.js");
function getNoMatchingSuggestion(valueType = "Any") {
  return {
    trailing: "",
    valueType,
    displayValue: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_react_intl.FormattedMessage,
      {
        ...import_messages.filterFieldMessages.suggestions.noMatchingSuggestions
      }
    )
  };
}
function getStartTypingSuggestion(suggestionType, valueType = "Any") {
  return {
    trailing: "",
    valueType,
    suggestionType: "hint-startTyping",
    displayValue: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_react_intl.FormattedMessage,
      {
        ...import_messages.filterFieldMessages.suggestions.startTyping,
        values: {
          suggestionType
        }
      }
    )
  };
}
function getPastingSuggestion(valueType = "Any") {
  return {
    trailing: "",
    valueType,
    suggestionType: "pastedContent",
    id: "strato-filterfield-suggestion-pasted-content",
    icon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.PasteIcon, {}),
    displayValue: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_intl.FormattedMessage, { ...import_messages.filterFieldMessages.suggestions.pasteRaw }),
    "data-strato-filter-field-should-escape-suggestion": true
  };
}
function getSearchSuggestion(value) {
  if (value === void 0) {
    return {
      value: '* ~ "',
      trailing: '"',
      cursorBeforeTrailing: true,
      id: "strato-filterfield-suggestion-search",
      displayValue: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_react_intl.FormattedMessage,
        {
          ...import_messages.filterFieldMessages.suggestions.search,
          values: {
            value: false
          }
        }
      ),
      icon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.MagnifyingGlassIcon, {}),
      insertionStrategy: "replace-statement"
    };
  }
  const match = /^"(.+)(?<!\\)" ?$/.exec(value);
  const textValue = match ? match[1] : value;
  const escapedSuggestion = (0, import_escape_suggestion.escapeSuggestion)(match ? value.trim() : value);
  return {
    value: `* ~ ${escapedSuggestion}`,
    id: "strato-filterfield-suggestion-search",
    displayValue: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_react_intl.FormattedMessage,
      {
        ...import_messages.filterFieldMessages.suggestions.search,
        values: {
          value: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
            "\xA0",
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)("mark", { children: textValue })
          ] })
        }
      }
    ),
    icon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.MagnifyingGlassIcon, {}),
    insertionStrategy: "replace-statement"
  };
}
function getFallbackKeySuggestion(key, operator, value) {
  const match = /^"(.+)(?<!\\)" ?$/.exec(value);
  const textValue = match ? match[1] : value;
  const escapedSuggestion = (0, import_escape_suggestion.escapeSuggestion)(match ? value.trim() : value);
  const baseOperator = operator.includes("not") ? "!=" : "=";
  switch (operator) {
    case "in":
    case "not in": {
      return {
        value: `${key} ${operator} (${escapedSuggestion})`,
        displayValue: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("span", { children: [
          key,
          " ",
          operator,
          " (",
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)("mark", { children: textValue }),
          ")"
        ] }),
        insertionStrategy: "replace-statement"
      };
    }
    case "contains":
    case "not-contains": {
      return {
        value: `${key} ${baseOperator} *${escapedSuggestion}*`,
        displayValue: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("span", { children: [
          key,
          " ",
          baseOperator,
          " *",
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)("mark", { children: textValue }),
          "*"
        ] }),
        insertionStrategy: "replace-statement"
      };
    }
    case "starts-with":
    case "not-starts-with": {
      return {
        value: `${key} ${baseOperator} ${escapedSuggestion}*`,
        displayValue: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("span", { children: [
          key,
          " ",
          baseOperator,
          " ",
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)("mark", { children: textValue }),
          "*"
        ] }),
        insertionStrategy: "replace-statement"
      };
    }
    case "ends-with":
    case "not-ends-with": {
      return {
        value: `${key} ${baseOperator} *${escapedSuggestion}`,
        displayValue: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("span", { children: [
          key,
          " ",
          baseOperator,
          " *",
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)("mark", { children: textValue })
        ] }),
        insertionStrategy: "replace-statement"
      };
    }
    default: {
      const operatorName = (0, import_operator_specification.getOperatorShortName)(operator);
      return {
        value: `${key} ${operatorName} ${escapedSuggestion}`,
        displayValue: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("span", { children: [
          key,
          " ",
          operatorName,
          " ",
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)("mark", { children: textValue })
        ] }),
        insertionStrategy: "replace-statement"
      };
    }
  }
}
function getStatementSuggestionForMatchingValue(key, value) {
  const match = /^"(.+)(?<!\\)" ?$/.exec(value);
  const textValue = match ? match[1] : value;
  const escapedSuggestion = (0, import_escape_suggestion.escapeSuggestion)(match ? value.trim() : value);
  return {
    value: `${key} = ${escapedSuggestion}`,
    // Add the class only to the value part as we only want to show highlighting in the value and not the key
    displayValue: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("span", { children: [
      key,
      " =",
      " ",
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { className: suggestionStyles.statementSuggestionValue, children: textValue })
    ] }),
    insertionStrategy: "replace-statement"
  };
}
function getJSONPathHint() {
  return {
    trailing: "",
    id: "strato-filterfield-suggestion-jsonpath-hint",
    suggestionType: "JSONPath",
    displayValue: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("span", { className: styles.suggestionHintContainer, children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_react_intl.FormattedMessage,
        {
          defaultMessage: "Filter JSON attributes using dot notation. For example: ",
          description: "Hint for entering JSONPath syntax in the FilterField.",
          id: "Ua7PnIRWAcErpii9"
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("span", { className: styles.filterStatementBase, children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { className: styles.filterKey, children: "content$.key[1].key " }),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { className: styles.operator, children: "=" }),
        " ",
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { className: styles.filterValue, children: "value" })
      ] })
    ] })
  };
}
