// @dynatrace/strato-components/filters/filter-field/utils/suggestions-utils/suggestion-type.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var suggestion_type_exports = {};
__export(suggestion_type_exports, {
  getValidValueTypes: () => getValidValueTypes
});
module.exports = __toCommonJS(suggestion_type_exports);
var import_type_guards = require("../validation-utils/type-guards.js");
var import_type_validators = require("../validation-utils/type-validators.js");
function getValidValueTypes(valuePredicate) {
  let valueTypes;
  let hasLiteralValues = false;
  if (typeof valuePredicate === "function") {
    valueTypes = ["Any"];
  } else {
    ({ types: valueTypes, hasLiteralValues } = getTypesFromValuePredicate(valuePredicate));
  }
  valueTypes = Array.from(new Set(valueTypes));
  const suggestionType = getSuggestionTypesFromPrimitiveArray(valueTypes);
  return { types: valueTypes, suggestionType, hasLiteralValues };
}
function getSuggestionTypeForStringSuggestions(valuePredicate) {
  switch (true) {
    case (0, import_type_validators.isValidIPAddress)(valuePredicate):
      return { types: ["IPAddress"], hasLiteralValues: true };
    case (0, import_type_validators.isValidTimestamp)(valuePredicate):
      return { types: ["Timestamp"], hasLiteralValues: true };
    case (0, import_type_validators.isValidUID)(valuePredicate):
      return { types: ["UID"], hasLiteralValues: true };
    case (0, import_type_validators.isValidSmartscapeId)(valuePredicate):
      return { types: ["SmartscapeId"], hasLiteralValues: true };
    default:
      return { types: ["String"], hasLiteralValues: true };
  }
}
function getTypesFromValuePredicate(valuePredicate) {
  if (valuePredicate === void 0) {
    return { types: ["Any"], hasLiteralValues: false };
  }
  if ((0, import_type_guards.isFilterFieldRichValuePredicate)(valuePredicate)) {
    return getTypesFromValuePredicate(valuePredicate.value);
  }
  if (typeof valuePredicate === "string") {
    return getSuggestionTypeForStringSuggestions(valuePredicate);
  }
  if (typeof valuePredicate === "number") {
    return { types: ["Number"], hasLiteralValues: true };
  }
  if (typeof valuePredicate === "boolean") {
    return { types: ["Boolean"], hasLiteralValues: true };
  }
  if ((0, import_type_guards.isFilterFieldValidatorMapDurationSpecification)(valuePredicate)) {
    return { types: ["Duration"], hasLiteralValues: true };
  }
  if ((0, import_type_guards.isFilterFieldValidatorMapTypeSpecification)(valuePredicate)) {
    const predicate = valuePredicate.type;
    if (Array.isArray(predicate)) {
      return {
        types: predicate.length > 0 ? Array.from(new Set(predicate)) : ["Any"],
        hasLiteralValues: false
      };
    }
    return { types: [predicate], hasLiteralValues: false };
  }
  if ((0, import_type_guards.isFilterFieldValueSuggestionGroupConfig)(valuePredicate)) {
    return { types: ["Any"], hasLiteralValues: true };
  }
  return getTypesFromPredicateArray(valuePredicate);
}
function getTypesFromPredicateArray(valuePredicate) {
  const types = [];
  let hasLiteralValues = false;
  for (const predicate of valuePredicate) {
    if ((0, import_type_guards.isFilterFieldValidatorMapTypeSpecification)(predicate)) {
      if (Array.isArray(predicate.type)) {
        types.push(...predicate.type);
      } else {
        types.push(predicate.type);
      }
    } else {
      hasLiteralValues = true;
    }
  }
  return {
    types: types.length > 0 ? Array.from(new Set(types)) : ["Any"],
    hasLiteralValues
  };
}
function getSuggestionTypesFromPrimitiveArray(primitives) {
  if (primitives.length === 1) {
    return primitives[0];
  }
  if (primitives.includes("JSONPath")) {
    if (primitives.length <= 2) {
      return primitives.sort((a, b) => a === "JSONPath" ? 1 : -1);
    }
    return ["Any", "JSONPath"];
  }
  return "Any";
}
