// @dynatrace/strato-components/filters/filter-field/utils/suggestions-utils/value-suggestions.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var value_suggestions_exports = {};
__export(value_suggestions_exports, {
  VALUE_GROUP_ID: () => VALUE_GROUP_ID,
  getDefaultValueOrWildcardSuggestion: () => getDefaultValueOrWildcardSuggestion,
  getValueSuggestionGroups: () => getValueSuggestionGroups
});
module.exports = __toCommonJS(value_suggestions_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_intl = require("react-intl");
var import_operator_suggestions = require("./operator-suggestions.js");
var import_shared_suggestions = require("./shared-suggestions.js");
var import_suggestion_group_utils = require("./suggestion-group-utils.js");
var import_suggestion_type = require("./suggestion-type.js");
var import_messages = require("../../messages.js");
var import_tree_node_guards = require("../tree-node-guards.js");
var import_type_guards = require("../validation-utils/type-guards.js");
var import_validation_helper_functions = require("../validation-utils/validation-helper-functions.js");
const VALUE_GROUP_ID = "strato-filter-field-value-suggestions";
function getValueSuggestionGroups(key, comparisonOperator, internalValidatorMap, currentToken, showDefaultValueOrWildcardSuggestions, showWildcardValueSuggestions, isPasting) {
  if (!isPasting && (!key || !internalValidatorMap) && !showDefaultValueOrWildcardSuggestions && !showWildcardValueSuggestions) {
    if (comparisonOperator === "search" && (currentToken.textValue === "" || currentToken.textValue === '""')) {
      return wrapSuggestionsInGroup((0, import_shared_suggestions.getStartTypingSuggestion)("search phrase"));
    }
    return currentToken.textValue ? [] : wrapSuggestionsInGroup((0, import_shared_suggestions.getStartTypingSuggestion)("value"));
  }
  if (!(0, import_tree_node_guards.isFilterFieldLeafNode)(currentToken)) {
    return [];
  }
  if (showDefaultValueOrWildcardSuggestions) {
    return wrapSuggestionsInGroup(
      ...getDefaultValueOrWildcardSuggestion(
        currentToken.textValue,
        currentToken.value,
        internalValidatorMap,
        key,
        comparisonOperator
      )
    );
  }
  if (showWildcardValueSuggestions) {
    return wrapSuggestionsInGroup(
      ...getValueWildcardOperatorSuggestions(
        key,
        internalValidatorMap,
        !!comparisonOperator
      )
    );
  }
  if (!key || !internalValidatorMap) {
    return wrapSuggestionsInGroup(
      getDefaultValueSuggestion(currentToken.textValue)
    );
  }
  const predicate = internalValidatorMap.normalizedKeys.byKey.get(key);
  if (!predicate?.valuePredicate || typeof predicate.valuePredicate === "function") {
    return currentToken.textValue ? wrapSuggestionsInGroup(
      getDefaultValueSuggestion(currentToken.textValue)
    ) : wrapSuggestionsInGroup((0, import_shared_suggestions.getStartTypingSuggestion)("value"));
  }
  const valuePredicateItems = Array.isArray(predicate.valuePredicate) ? predicate.valuePredicate : [predicate.valuePredicate];
  const groups = [];
  let pendingItems = [];
  function processPendingItems() {
    if (pendingItems.length === 0) {
      return;
    }
    groups.push({
      id: `${VALUE_GROUP_ID}-implicit-${groups.length}`,
      suggestions: pendingItems
    });
    pendingItems = [];
  }
  for (let index = 0; index < valuePredicateItems.length; index++) {
    const item = valuePredicateItems[index];
    if ((0, import_type_guards.isFilterFieldValueSuggestionGroupConfig)(item)) {
      processPendingItems();
      const groupSuggestions = [];
      for (const suggestion of item.suggestions) {
        collectItem(
          getSuggestionFromValuePredicate(suggestion, currentToken.textValue),
          groupSuggestions
        );
      }
      if (groupSuggestions.length > 0) {
        groups.push({
          id: (0, import_suggestion_group_utils.resolveGroupId)(item, `value-group-${index}`),
          label: item.label,
          "aria-label": item["aria-label"],
          suggestions: groupSuggestions
        });
      }
      continue;
    }
    collectItem(
      getSuggestionFromValuePredicate(item, currentToken.textValue),
      pendingItems
    );
  }
  processPendingItems();
  const suggestionMatchesTextValue = groups.some(
    (group) => group.suggestions.some(
      (suggestion) => suggestion.suggestionType === "validatorMap" && suggestion.value === currentToken.textValue
    )
  );
  const result = [];
  for (const group of groups) {
    const suggestions = suggestionMatchesTextValue || !currentToken.textValue && group.suggestions.length > 1 ? group.suggestions.filter((s) => s.suggestionType === "validatorMap") : group.suggestions;
    if (suggestions.length > 0) {
      result.push({ ...group, suggestions });
    }
  }
  return result;
}
function wrapSuggestionsInGroup(...suggestions) {
  return suggestions.length > 0 ? [{ id: VALUE_GROUP_ID, suggestions }] : [];
}
function getDefaultValueOrWildcardSuggestion(textValue, value, internalValidatorMap, keyValue, comparisonOperator) {
  const comparisonOperators = (0, import_validation_helper_functions.getValidOperators)(keyValue, internalValidatorMap);
  const isValidOperator = comparisonOperator && comparisonOperators.includes(comparisonOperator);
  switch (comparisonOperator) {
    case "starts-with":
      return isValidOperator ? [getStartsWithValueSuggestion(textValue, value)] : [getStartsWithIsNotSupportedSuggestion()];
    case "ends-with":
      return isValidOperator ? [getEndsWithValueSuggestion(textValue, value)] : [getEndsWithIsNotSupportedSuggestion()];
    case "contains":
      return isValidOperator ? [getContainsValueSuggestion(textValue, value)] : [getContainsIsNotSupportedSuggestion()];
    case "not-starts-with":
      return isValidOperator ? [getNotStartsWithValueSuggestion(textValue, value)] : [getStartsWithIsNotSupportedSuggestion()];
    case "not-ends-with":
      return isValidOperator ? [getNotEndsWithValueSuggestion(textValue, value)] : [getEndsWithIsNotSupportedSuggestion()];
    case "not-contains":
      return isValidOperator ? [getNotContainsValueSuggestion(textValue, value)] : [getContainsIsNotSupportedSuggestion()];
    default: {
      if (textValue && typeof comparisonOperator !== "boolean") {
        return [getDefaultValueSuggestion(textValue)];
      }
      return [(0, import_shared_suggestions.getStartTypingSuggestion)("value")];
    }
  }
}
function collectItem(result, suggestions) {
  if (result === void 0) {
    return;
  }
  if (Array.isArray(result)) {
    suggestions.push(...result);
    return;
  }
  suggestions.push(result);
}
function getSuggestionFromValuePredicate(predicate, textValue) {
  const { suggestionType } = (0, import_suggestion_type.getValidValueTypes)(predicate);
  const predicateValue = getPredicateValue(predicate);
  if (!Array.isArray(predicate) && (typeof predicate !== "object" || (0, import_type_guards.isFilterFieldValidatorMapDurationSpecification)(predicate) || (0, import_type_guards.isFilterFieldRichValuePredicate)(predicate))) {
    const richPredicate = (0, import_type_guards.isFilterFieldRichValuePredicate)(predicate) ? predicate : void 0;
    return {
      value: predicateValue,
      displayValue: richPredicate?.displayValue,
      details: richPredicate?.details,
      suggestionType: "validatorMap",
      valueType: Array.isArray(suggestionType) ? "JSONPath" : suggestionType
    };
  }
  if ((0, import_type_guards.isFilterFieldValidatorMapTypeSpecification)(predicate) && predicate.type === "Boolean") {
    return [
      { value: "true", suggestionType: "validatorMap", valueType: "Boolean" },
      { value: "false", suggestionType: "validatorMap", valueType: "Boolean" }
    ];
  }
  if (textValue) {
    return getDefaultValueSuggestion(
      textValue,
      Array.isArray(suggestionType) ? "Any" : suggestionType
    );
  } else {
    return (0, import_shared_suggestions.getStartTypingSuggestion)(
      "value",
      Array.isArray(suggestionType) ? "Any" : suggestionType
    );
  }
}
function getPredicateValue(predicate) {
  if (Array.isArray(predicate)) {
    return "";
  }
  if (typeof predicate !== "object") {
    return predicate.toString();
  }
  if ((0, import_type_guards.isFilterFieldRichValuePredicate)(predicate)) {
    return `${predicate.value}`;
  }
  if ((0, import_type_guards.isFilterFieldValidatorMapDurationSpecification)(predicate)) {
    return `${predicate.value}${predicate.unit}`;
  }
  return "";
}
function getValueWildcardOperatorSuggestions(key, internalValidatorMap, currentOperatorValue) {
  const operators = getValidOperatorsShortForWildcardOperators(
    key,
    internalValidatorMap
  );
  return operators.filter((operator) => {
    if (currentOperatorValue) {
      return operator === "= *" || !operator.includes("not") && operator !== "!= *";
    }
    return operator === "!= *" || operator.includes("not");
  }).map(
    (operator) => import_operator_suggestions.operatorSuggestionsWithDetails[operator] ?? { value: operator }
  );
}
function getValidOperatorsShortForWildcardOperators(key, internalValidatorMap) {
  const operators = (0, import_validation_helper_functions.getValidOperators)(key, internalValidatorMap);
  const result = [];
  if (operators.includes("exists")) {
    result.push("= *");
  }
  if (operators.includes("not-exists")) {
    result.push("!= *");
  }
  if (operators.includes("contains")) {
    result.push("contains");
  }
  if (operators.includes("starts-with")) {
    result.push("starts-with");
  }
  if (operators.includes("ends-with")) {
    result.push("ends-with");
  }
  if (operators.includes("not-starts-with")) {
    result.push("not-starts-with");
  }
  if (operators.includes("not-contains")) {
    result.push("not-contains");
  }
  if (operators.includes("not-ends-with")) {
    result.push("not-ends-with");
  }
  return result;
}
function getDefaultValueSuggestion(value, valueType = "Any") {
  const match = /^"(.+)(?<!\\)"$/.exec(value);
  const textValue = match ? match[1] : value;
  return {
    trailing: " ",
    value,
    valueType,
    suggestionType: "hint-useAsValue",
    id: "strato-filterfield-suggestion-use-as-value",
    displayValue: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_react_intl.FormattedMessage,
      {
        ...import_messages.filterFieldMessages.suggestions.useAsValue,
        values: {
          value: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
            "\xA0",
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)("mark", { children: textValue }),
            "\xA0"
          ] })
        }
      }
    )
  };
}
function getStartsWithIsNotSupportedSuggestion(valueType = "Any") {
  return {
    trailing: "",
    valueType,
    suggestionType: "hint-wildcard",
    id: "strato-filterfield-suggestion-starts-with-not-supported",
    displayValue: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_react_intl.FormattedMessage,
      {
        ...import_messages.filterFieldMessages.unsupportedOperators.startsWith
      }
    )
  };
}
function getEndsWithIsNotSupportedSuggestion(valueType = "Any") {
  return {
    trailing: "",
    valueType,
    suggestionType: "hint-wildcard",
    id: "strato-filterfield-suggestion-ends-with-not-supported",
    displayValue: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_react_intl.FormattedMessage,
      {
        ...import_messages.filterFieldMessages.unsupportedOperators.endsWith
      }
    )
  };
}
function getContainsIsNotSupportedSuggestion(valueType = "Any") {
  return {
    trailing: "",
    valueType,
    suggestionType: "hint-wildcard",
    id: "strato-filterfield-suggestion-contains-not-supported",
    displayValue: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_react_intl.FormattedMessage,
      {
        ...import_messages.filterFieldMessages.unsupportedOperators.contains
      }
    )
  };
}
function getStartsWithValueSuggestion(textValue, value, valueType = "Any") {
  return {
    trailing: " ",
    value: textValue,
    valueType,
    suggestionType: "hint-wildcard",
    id: "strato-filterfield-suggestion-starts-with-value",
    displayValue: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_react_intl.FormattedMessage,
      {
        ...import_messages.filterFieldMessages.value.startsWith,
        values: {
          value: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("mark", { children: value })
        }
      }
    ),
    escapeSuggestion: false
  };
}
function getEndsWithValueSuggestion(textValue, value, valueType = "Any") {
  return {
    trailing: " ",
    value: textValue,
    valueType,
    suggestionType: "hint-wildcard",
    id: "strato-filterfield-suggestion-ends-with-value",
    displayValue: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_react_intl.FormattedMessage,
      {
        ...import_messages.filterFieldMessages.value.endsWith,
        values: {
          value: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("mark", { children: value })
        }
      }
    ),
    escapeSuggestion: false
  };
}
function getContainsValueSuggestion(textValue, value, valueType = "Any") {
  return {
    trailing: " ",
    value: textValue,
    valueType,
    suggestionType: "hint-wildcard",
    id: "strato-filterfield-suggestion-contains-value",
    displayValue: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_react_intl.FormattedMessage,
      {
        ...import_messages.filterFieldMessages.value.contains,
        values: {
          value: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("mark", { children: value })
        }
      }
    ),
    escapeSuggestion: false
  };
}
function getNotStartsWithValueSuggestion(textValue, value, valueType = "Any") {
  return {
    trailing: " ",
    value: textValue,
    valueType,
    suggestionType: "hint-wildcard",
    id: "strato-filterfield-suggestion-not-starts-with-value",
    displayValue: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_react_intl.FormattedMessage,
      {
        ...import_messages.filterFieldMessages.value.notStartsWith,
        values: {
          value: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("mark", { children: value })
        }
      }
    ),
    escapeSuggestion: false
  };
}
function getNotEndsWithValueSuggestion(textValue, value, valueType = "Any") {
  return {
    trailing: " ",
    value: textValue,
    valueType,
    suggestionType: "hint-wildcard",
    id: "strato-filterfield-suggestion-not-ends-with-value",
    displayValue: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_react_intl.FormattedMessage,
      {
        ...import_messages.filterFieldMessages.value.notEndsWith,
        values: {
          value: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("mark", { children: value })
        }
      }
    ),
    escapeSuggestion: false
  };
}
function getNotContainsValueSuggestion(textValue, value, valueType = "Any") {
  return {
    trailing: " ",
    value: textValue,
    valueType,
    suggestionType: "hint-wildcard",
    id: "strato-filterfield-suggestion-not-contains-value",
    displayValue: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_react_intl.FormattedMessage,
      {
        ...import_messages.filterFieldMessages.value.notContains,
        values: {
          value: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("mark", { children: value })
        }
      }
    ),
    escapeSuggestion: false
  };
}
