// @dynatrace/strato-components/charts/tree-map/store/selectors.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var selectors_exports = {};
__export(selectors_exports, {
  useClusterNodesState: () => useClusterNodesState,
  useColorMap: () => useColorMap,
  useGroupsSelectedState: () => useGroupsSelectedState,
  useHighlightedState: () => useHighlightedState,
  useLeafNodesState: () => useLeafNodesState,
  useSelector: () => useSelector,
  useSetState: () => useSetState,
  useTreeMapState: () => useTreeMapState,
  useTreeMapToolbarVisibleState: () => useTreeMapToolbarVisibleState
});
module.exports = __toCommonJS(selectors_exports);
var import_react = require("react");
var import_store = require("./store.js");
var import_selector = require("../../../core/store/selector.js");
const useSelector = (selector) => (0, import_selector.useStoreSelector)(import_store.TreeMapStateContext, selector);
const useSetState = () => {
  const { setState } = (0, import_react.useContext)(import_store.TreeMapStateContext);
  return setState;
};
const useTreeMapState = () => useSelector((state) => state);
const useHighlightedState = () => useSelector((state) => state.highlightedItem);
const useGroupsSelectedState = () => useSelector((state) => state.groupsSelected);
const useLeafNodesState = () => useSelector((state) => state.hierarchyLeafNodes);
const useClusterNodesState = () => useSelector((state) => state.hierarchyClusterNodes);
const useTreeMapToolbarVisibleState = () => useSelector((state) => state.overlay.toolbarVisible);
const useColorMap = () => useSelector((state) => state.colorMap);
