// @dynatrace/strato-components/charts/tree-map/store/store.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var store_exports = {};
__export(store_exports, {
  TreeMapStateContext: () => TreeMapStateContext,
  initialTreeMapState: () => initialTreeMapState,
  useSetState: () => useSetState
});
module.exports = __toCommonJS(store_exports);
var import_react = require("react");
var import_store = require("../../../core/store/store.js");
const initialTreeMapState = {
  groupsSelected: [],
  highlightedItem: void 0,
  overlay: {
    toolbarVisible: false
  },
  hierarchyLeafNodes: [],
  hierarchyClusterNodes: [],
  colorMap: /* @__PURE__ */ new Map()
};
const TreeMapStateContext = (0, import_react.createContext)(
  (0, import_store.createStore)(initialTreeMapState)
);
const useSetState = () => {
  const { setState } = (0, import_react.useContext)(TreeMapStateContext);
  return setState;
};
