// @dynatrace/strato-components/esm/charts/histogram/components/overlays/HistogramDatapoints.js
import { jsx } from "react/jsx-runtime";
import { memo } from "react";
import { datapointLayout } from "./HistogramOverlays.sty.js";
import { Canvas } from "../../../core/components/canvas/Canvas.js";
import { CanvasContext } from "../../../core/components/canvas/CanvasContext.js";
import { useGridPlotSize } from "../../../core/components/chart-grid/hooks/useGridPlotSize.js";
import { OVERLAY_PADDING } from "../../../core/components/data-point/constants.js";
import { DataPointCanvas } from "../../../core/components/data-point/DataPointCanvas.js";
const HistogramDatapoints = memo(
  (props) => {
    const { closestPoint, groupedPoints } = props;
    const { width, height } = useGridPlotSize();
    return /* @__PURE__ */ jsx("div", { className: datapointLayout, children: /* @__PURE__ */ jsx(
      CanvasContext.Provider,
      {
        value: {
          shapeRefs: /* @__PURE__ */ new Map()
        },
        children: /* @__PURE__ */ jsx(
          Canvas,
          {
            width: width + OVERLAY_PADDING * 2,
            height: height + OVERLAY_PADDING * 2,
            children: groupedPoints.length > 0 && groupedPoints.map((pointGroup, index) => /* @__PURE__ */ jsx(
              DataPointCanvas,
              {
                points: pointGroup,
                closestPoint
              },
              `pointGroup-${index}`
            ))
          }
        )
      }
    ) });
  }
);
HistogramDatapoints["displayName"] = "HistogramDatapoints";
export {
  HistogramDatapoints
};
