// @dynatrace/strato-components/charts/timeseries/utils/chart-configuration.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var chart_configuration_exports = {};
__export(chart_configuration_exports, {
  createTimeseriesConfiguration: () => createTimeseriesConfiguration,
  deserializeConfig: () => deserializeConfig,
  getColorPaletteConfiguration: () => getColorPaletteConfiguration,
  getCurrentZoomConfiguration: () => getCurrentZoomConfiguration,
  getGapPolicyConfiguration: () => getGapPolicyConfiguration,
  getInfiniteZoomConfiguration: () => getInfiniteZoomConfiguration,
  getInitialZoomConfiguration: () => getInitialZoomConfiguration,
  getLegendConfiguration: () => getLegendConfiguration,
  getPointsDisplayConfiguration: () => getPointsDisplayConfiguration,
  getSeriesCurveConfiguration: () => getSeriesCurveConfiguration,
  getThresholdsConfiguration: () => getThresholdsConfiguration,
  getToolbar: () => getToolbar,
  getTooltipConfiguration: () => getTooltipConfiguration,
  getValueRepresentationConfiguration: () => getValueRepresentationConfiguration,
  getXAxisConfiguration: () => getXAxisConfiguration,
  getYAxisConfiguration: () => getYAxisConfiguration,
  normalizeConfig: () => normalizeConfig
});
module.exports = __toCommonJS(chart_configuration_exports);
var import_lodash_es = require("lodash");
var import_is_valid_points_display = require("../../core/components/canvas/utils/is-valid-points-display.js");
var import_constants = require("../../core/constants.js");
var import_ColorRule = require("../../core/slots/ColorRule.js");
var import_color_palette = require("../../core/types/color-palette.js");
var import_getValidColor = require("../../core/utils/getValidColor.js");
var import_interactions = require("../../core/utils/interactions.js");
var import_is_valid_color_palette = require("../../core/utils/is-valid-color-palette.js");
var import_is_valid_legend_position = require("../../core/utils/is-valid-legend-position.js");
var import_is_valid_series_display_mode = require("../../core/utils/is-valid-series-display-mode.js");
var import_is_valid_tooltip_variant = require("../../core/utils/is-valid-tooltip-variant.js");
var import_is_valid_value_representation = require("../../core/utils/is-valid-value-representation.js");
var import_constants2 = require("../constants.js");
const sanitizeYAxisConfig = (yAxes) => {
  const reversedYAxes = yAxes.reverse();
  const leftAxis = reversedYAxes.find(
    (axis) => axis?.position === "left" || axis?.position === void 0
  );
  const rightAxis = reversedYAxes.find((axis) => axis?.position === "right");
  const sanitizedYAxes = {};
  if (leftAxis) {
    sanitizedYAxes.left = leftAxis;
  }
  if (rightAxis) {
    sanitizedYAxes.right = rightAxis;
  }
  return sanitizedYAxes;
};
function getYAxisConfiguration(slotsYAxis, contextYAxis) {
  const contextConfig = contextYAxis === void 0 ? {} : sanitizeYAxisConfig(contextYAxis);
  const slotsConfig = slotsYAxis === void 0 ? {} : sanitizeYAxisConfig(slotsYAxis);
  const { left, right } = (0, import_lodash_es.merge)({}, contextConfig, slotsConfig);
  if (left === void 0 && right === void 0) {
    return import_constants2.defaultYAxis;
  }
  const result = {};
  if (left) {
    result.left = (0, import_lodash_es.merge)({}, import_constants2.defaultYAxis.left, left);
  }
  if (right) {
    const defaultRight = { ...import_constants2.defaultYAxis.left, position: "right" };
    result.right = (0, import_lodash_es.merge)({}, defaultRight, right);
  }
  return result;
}
function getLegendConfiguration(contextLegend, slotsLegend) {
  const legendConfig = (0, import_lodash_es.merge)(
    {},
    import_constants2.defaultLegend,
    contextLegend,
    slotsLegend
  );
  legendConfig.position = (0, import_is_valid_legend_position.isValidLegendPosition)(legendConfig.position) ? legendConfig.position : import_constants2.defaultLegend.position;
  return legendConfig;
}
function getTooltipConfiguration(contextTooltip, slotsTooltip) {
  const tooltip = (0, import_lodash_es.merge)(
    {},
    { variant: import_constants2.defaultTooltipVariant },
    contextTooltip,
    slotsTooltip
  );
  tooltip.variant = (0, import_is_valid_tooltip_variant.isValidTooltipVariant)(tooltip.variant) ? tooltip.variant : import_constants2.defaultTooltipVariant;
  if (tooltip.seriesDisplayMode) {
    tooltip.seriesDisplayMode = (0, import_is_valid_series_display_mode.isValidSeriesDisplayMode)(
      tooltip.seriesDisplayMode
    ) ? tooltip.seriesDisplayMode : import_constants2.defaultSeriesDisplayMode;
  } else {
    tooltip.seriesDisplayMode = tooltip.variant === "shared" ? "single-line" : "multi-line";
  }
  return tooltip;
}
function getXAxisConfiguration(contextXAxis, slotsXAxis) {
  return (0, import_lodash_es.merge)({}, contextXAxis, slotsXAxis);
}
function getThresholdsConfiguration(contextThresholds = [], slotsThresholds = []) {
  const allThresholds = contextThresholds.concat(slotsThresholds).map((threshold) => ({
    ...threshold,
    id: crypto.randomUUID(),
    position: !threshold.position ? "left" : threshold.position,
    color: (0, import_getValidColor.isValidCSSColor)(threshold.color) ? threshold.color : import_constants2.DEFAULT_SCALE_FALLBACK_COLOR
  }));
  if (!allThresholds.length) {
    return void 0;
  }
  const thresholdConfig = {};
  thresholdConfig.left = allThresholds.filter(
    (threshold) => threshold.position === "left"
  );
  thresholdConfig.right = allThresholds.filter(
    (threshold) => threshold.position === "right"
  );
  return thresholdConfig;
}
const isUndefinedOrIsEmpty = (value) => {
  return (0, import_lodash_es.isUndefined)(value) && (0, import_lodash_es.isEmpty)(value);
};
const processToolbarConfig = (config) => {
  if (!config) {
    return config;
  }
  const { hidden, ...rest } = config;
  if (hidden === true) {
    return { ...rest, enabled: false };
  }
  return config;
};
function getToolbar(contextToolbar, slotsChartToolbar, slotDownloadCSV, contextDownloadCSV) {
  const hasNoDownloadData = isUndefinedOrIsEmpty(slotDownloadCSV) && isUndefinedOrIsEmpty(contextDownloadCSV);
  if (!contextToolbar && (0, import_lodash_es.isEmpty)(slotsChartToolbar) && hasNoDownloadData) {
    return { ...import_constants2.defaultToolbar };
  }
  const processedContextToolbar = processToolbarConfig(contextToolbar);
  const processedSlotsToolbar = processToolbarConfig(slotsChartToolbar);
  return (0, import_lodash_es.merge)(
    {},
    import_constants2.defaultToolbar,
    processedContextToolbar,
    processedSlotsToolbar,
    {
      downloadData: slotDownloadCSV ?? contextDownloadCSV
    }
  );
}
function getGapPolicyConfiguration(contextGapPolicy, propsGapPolicy) {
  return propsGapPolicy ?? contextGapPolicy ?? import_constants2.defaultGapPolicy;
}
function getPointsDisplayConfiguration(contextPointsDisplay, propsPointsDisplay) {
  const pointsDisplay = propsPointsDisplay ?? contextPointsDisplay ?? import_constants2.defaultPointsDisplay;
  return (0, import_is_valid_points_display.isValidPointsDisplay)(pointsDisplay) ? pointsDisplay : import_constants2.defaultPointsDisplay;
}
function getValueRepresentationConfiguration(contextValueRepresentation, propsValueRepresentation) {
  const valueRepresentation = propsValueRepresentation ?? contextValueRepresentation ?? import_constants2.defaultValueRepresentation;
  return (0, import_is_valid_value_representation.isValidValueRepresentation)(valueRepresentation) ? valueRepresentation : import_constants2.defaultValueRepresentation;
}
function getColorPaletteConfiguration(contextColorPalette, propsColorPalette) {
  const colorPalette = propsColorPalette ?? contextColorPalette ?? import_constants2.defaultColorPalette;
  return (0, import_color_palette.isCustomColorPalette)(colorPalette) || (0, import_is_valid_color_palette.isValidColorPalette)(colorPalette) ? colorPalette : import_constants2.defaultTimeseriesChartConfig.colorPalette ?? import_constants2.defaultColorPalette;
}
function getSeriesCurveConfiguration(contextConfig, propsConfig) {
  return propsConfig ?? contextConfig ?? import_constants2.defaultSeriesCurve;
}
function getInfiniteZoomConfiguration(contextConfig, propsConfig) {
  return propsConfig ?? contextConfig ?? import_constants.DEFAULT_INFINITE_ZOOM;
}
function getInitialZoomConfiguration(contextConfig, propsConfig) {
  return propsConfig ?? contextConfig;
}
function getCurrentZoomConfiguration(contextConfig, propsConfig) {
  return propsConfig ?? contextConfig;
}
function createTimeseriesConfiguration(slots, contextConfig, timeseriesChartConfigProps) {
  const yAxis = getYAxisConfiguration(slots.yAxis, contextConfig.yAxis);
  const truncationMode = timeseriesChartConfigProps?.truncationMode ?? contextConfig.truncationMode ?? import_constants.DEFAULT_TRUNCATION_MODE;
  const legend = getLegendConfiguration(contextConfig.legend, slots.legend);
  const tooltip = getTooltipConfiguration(contextConfig.tooltip, slots.tooltip);
  const xAxis = getXAxisConfiguration(contextConfig.xAxis, slots.xAxis);
  const configThresholds = contextConfig.thresholdIndicators || [];
  const thresholds = getThresholdsConfiguration(
    configThresholds,
    slots.thresholds
  );
  const toolbar = getToolbar(contextConfig.toolbar, slots.toolbar);
  const propsInteractions = {
    onZoomChange: timeseriesChartConfigProps.onZoomChange
  };
  const hasContextInteractions = contextConfig && (contextConfig.pan || contextConfig.zoom);
  const chartConfigInteractions = hasContextInteractions ? {
    pan: { enabled: contextConfig.pan?.disabled !== true },
    zoom: { enabled: contextConfig.zoom?.disabled !== true },
    zoomX: { enabled: contextConfig.zoom?.disabled !== true }
  } : {};
  const interactions = (0, import_interactions.getInteractions)(
    contextConfig.interactions,
    slots.interactions,
    propsInteractions,
    chartConfigInteractions
  );
  const gapPolicy = getGapPolicyConfiguration(
    contextConfig.gapPolicy,
    timeseriesChartConfigProps?.gapPolicy
  );
  const pointsDisplay = getPointsDisplayConfiguration(
    contextConfig.pointsDisplay,
    timeseriesChartConfigProps?.pointsDisplay
  );
  const valueRepresentation = getValueRepresentationConfiguration(
    contextConfig.valueRepresentation,
    timeseriesChartConfigProps?.valueRepresentation
  );
  const colorPalette = getColorPaletteConfiguration(
    contextConfig.colorPalette,
    timeseriesChartConfigProps?.colorPalette
  );
  const curve = getSeriesCurveConfiguration(
    contextConfig.curve,
    timeseriesChartConfigProps?.curve
  );
  const infiniteZoom = getInfiniteZoomConfiguration(
    contextConfig.infiniteZoom,
    timeseriesChartConfigProps.infiniteZoom
  );
  const initialZoom = getInitialZoomConfiguration(
    contextConfig.initialZoom,
    timeseriesChartConfigProps.initialZoom
  );
  const currentZoom = getCurrentZoomConfiguration(
    contextConfig.currentZoom,
    timeseriesChartConfigProps.currentZoom
  );
  const select = timeseriesChartConfigProps.select ?? slots.select;
  const colorRules = (contextConfig.colorRules ?? []).filter(import_ColorRule.isValidColorRule).concat(slots.colorRules ?? []);
  return {
    gapPolicy,
    pointsDisplay,
    valueRepresentation,
    colorPalette,
    legend,
    tooltip,
    toolbar,
    xAxis,
    yAxis,
    annotations: slots.annotations,
    truncationMode,
    thresholds,
    interactions,
    initialZoom,
    currentZoom,
    infiniteZoom,
    curve,
    select,
    colorRules
  };
}
function parseYAxisConfig(yAxis) {
  const { left, right } = yAxis;
  const yAxisConfig = [];
  if (left) {
    yAxisConfig.push(left);
  }
  if (right) {
    yAxisConfig.push(right);
  }
  return yAxisConfig;
}
function parseThresholdsConfig(thresholds) {
  const { left, right } = thresholds;
  const thresholdsConfig = [];
  if (left) {
    thresholdsConfig.push(...left);
  }
  if (right) {
    thresholdsConfig.push(...right);
  }
  return thresholdsConfig;
}
function normalizeConfig(config) {
  const { thresholds, ...rest } = config;
  const thresholdIndicators = thresholds ? parseThresholdsConfig(thresholds) : void 0;
  return {
    ...rest,
    yAxis: parseYAxisConfig(config.yAxis),
    thresholdIndicators
  };
}
const dateFormat = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z$/;
function reviver(key, value) {
  if (typeof value === "string" && dateFormat.test(value)) {
    return new Date(value);
  }
  return value;
}
function deserializeConfig(serializedConfig) {
  try {
    return (0, import_lodash_es.merge)(
      {},
      import_constants2.defaultTimeseriesChartConfig,
      JSON.parse(serializedConfig, reviver)
    );
  } catch {
    return import_constants2.defaultTimeseriesChartConfig;
  }
}
