// @dynatrace/strato-components/charts/timeseries/utils/get-datapoint-value.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var get_datapoint_value_exports = {};
__export(get_datapoint_value_exports, {
  getDatapointKey: () => getDatapointKey,
  getDatapointValue: () => getDatapointValue
});
module.exports = __toCommonJS(get_datapoint_value_exports);
var import_timeseries = require("./timeseries.js");
var import_get_relative_value = require("../../core/utils/get-relative-value.js");
var import_get_unit_as_string = require("../../core/utils/get-unit-as-string.js");
var import_safe_get = require("../../core/utils/safe-get.js");
const getDatapointKey = (timestamp, unit) => `${timestamp}_${unit}`;
function getDatapointValue(datapoint, variant, unit, relativeGroupedValues) {
  if (variant === "band") {
    const { y0, y1 } = datapoint;
    if (relativeGroupedValues) {
      const seriesUnit = (0, import_get_unit_as_string.getUnitAsString)(unit);
      const { maxValuesByUnit } = relativeGroupedValues;
      const total = (0, import_safe_get.safeGet)(maxValuesByUnit, seriesUnit);
      const relativeY0Value = (0, import_get_relative_value.getRelativeValue)(y0, total);
      const relativeY1Value = (0, import_get_relative_value.getRelativeValue)(y1, total);
      return {
        y0: relativeY0Value,
        absoluteY0: y0,
        y1: relativeY1Value,
        absoluteY1: y1
      };
    }
    return { y0, y1 };
  }
  const { value, start, end, center } = datapoint;
  if (relativeGroupedValues) {
    const { groupedValues, maxValuesByUnit } = relativeGroupedValues;
    const seriesUnit = (0, import_get_unit_as_string.getUnitAsString)(unit);
    const datapointKey = getDatapointKey(
      (0, import_timeseries.getDatapointTime)(start, end, center),
      seriesUnit
    );
    const sumOfSeriesValues = (0, import_safe_get.safeGet)(groupedValues, datapointKey) ?? 0;
    const isTotalValuePerUnit = relativeGroupedValues.uniqueUnits.get(seriesUnit) || variant === "line";
    const total = isTotalValuePerUnit ? maxValuesByUnit.get(seriesUnit) : sumOfSeriesValues;
    const relativeValue = (0, import_get_relative_value.getRelativeValue)(value, total);
    return { absolute: value, value: relativeValue };
  }
  return { value };
}
