// @dynatrace/strato-components/charts/timeseries/utils/iterate-config-slots.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var iterate_config_slots_exports = {};
__export(iterate_config_slots_exports, {
  iterateChartSeriesSlots: () => iterateChartSeriesSlots,
  iterateConfigSlots: () => iterateConfigSlots
});
module.exports = __toCommonJS(iterate_config_slots_exports);
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_react_is = require("react-is");
var import_Intent = require("../../../core/slots/Intent/Intent.js");
var import_is_slot = require("../../../core/utils/is-slot.js");
var import_populate_missing_marker_props = require("../../core/components/annotations/utils/populate-missing-marker-props.js");
var import_populate_missing_track_props = require("../../core/components/annotations/utils/populate-missing-track-props.js");
var import_ChartInteractions = require("../../core/slots/chart-interactions/ChartInteractions.js");
var import_ColorRule = require("../../core/slots/ColorRule.js");
var import_ChartToolbar = require("../../core/slots/toolbar/ChartToolbar.js");
var import_CompactToolbar = require("../../core/slots/toolbar/CompactToolbar.js");
var import_DownloadData = require("../../core/slots/toolbar/DownloadData.js");
var import_extract_behavior_tracking_props = require("../../core/utils/extract-behavior-tracking-props.js");
var import_interactions2 = require("../../core/utils/interactions.js");
var import_iterate_chart_interactions_slots = require("../../core/utils/iterate-chart-interactions-slots.js");
var import_Annotations = require("../slots/annotations/Annotations.js");
var import_Marker = require("../slots/annotations/Marker.js");
var import_Tooltip = require("../slots/annotations/Tooltip.js");
var import_Track = require("../slots/annotations/Track.js");
var import_DownloadCSV = require("../slots/DownloadCSV.js");
var import_Legend = require("../slots/Legend.js");
var import_Pan = require("../slots/Pan.js");
var import_Select = require("../slots/Select.js");
var import_Area = require("../slots/shapes/Area.js");
var import_Band = require("../slots/shapes/Band.js");
var import_Bar = require("../slots/shapes/Bar.js");
var import_Line = require("../slots/shapes/Line.js");
var import_ThresholdIndicator = require("../slots/ThresholdIndicator.js");
var import_Toolbar = require("../slots/Toolbar.js");
var import_Tooltip2 = require("../slots/Tooltip.js");
var import_XAxis = require("../slots/XAxis.js");
var import_YAxis = require("../slots/YAxis.js");
var import_Zoom = require("../slots/Zoom.js");
const slotInteractionsDefaultValues = {
  zoom: { enabled: true },
  zoomX: { enabled: true },
  pan: { enabled: true }
};
const chartInteractionsDefaultValues = {
  zoom: { enabled: false },
  zoomX: { enabled: false },
  pan: { enabled: false }
};
const getInteractionsConfig = (chartConfig, slotsConfig) => {
  const chartApiValues = (0, import_interactions2.getConfigPlusDefaultOrEmpty)(
    chartInteractionsDefaultValues,
    chartConfig
  );
  const slotsApiValues = (0, import_interactions2.getConfigPlusDefaultOrEmpty)(
    slotInteractionsDefaultValues,
    slotsConfig
  );
  return (0, import_lodash_es.merge)({}, chartApiValues, slotsApiValues);
};
const disabledOnlyWhenSpecified = ({ disabled }) => disabled;
const iterateConfigSlots = (children) => {
  let slots = {
    toolbar: {},
    legend: {},
    xAxis: {},
    tooltip: {},
    yAxis: [],
    thresholds: [],
    interactions: {},
    colorRules: []
  };
  let intents;
  let slotToolbar = {};
  let chartToolbar = {};
  let slotInteractions = {};
  let chartInteractions = {};
  import_react.Children.forEach(children, (child) => {
    if ((0, import_is_slot.isSlot)(child, import_ChartInteractions.ChartInteractions)) {
      const { children: children2, ...previousInteractions } = child.props;
      const interactions = (0, import_iterate_chart_interactions_slots.iterateChartInteractionsSlots)(children2);
      chartInteractions = {
        ...previousInteractions,
        ...interactions
      };
    } else if ((0, import_is_slot.isSlot)(child, import_ChartToolbar.ChartToolbar)) {
      const { children: children2, ...toolbar } = child.props;
      const childrenToolbar = iterateChartToolbarSlots(children2);
      chartToolbar = {
        enabled: true,
        ...toolbar,
        ...childrenToolbar
      };
    } else if ((0, import_is_slot.isSlot)(child, import_Toolbar.Toolbar)) {
      slotToolbar = {
        enabled: child.props.hidden !== true,
        ...slotToolbar,
        ...(0, import_CompactToolbar.sanitizeProps)(child.props),
        ...(0, import_extract_behavior_tracking_props.extractBehaviorTrackingProps)(child.props)
      };
    } else if ((0, import_is_slot.isSlot)(child, import_DownloadCSV.TimeseriesChartDownloadCSVSlot)) {
      slotToolbar = { ...slotToolbar, downloadData: { enabled: true } };
    } else if ((0, import_is_slot.isSlot)(child, import_Zoom.Zoom)) {
      const enabled = !disabledOnlyWhenSpecified(child.props);
      slotInteractions = {
        ...slotInteractions,
        zoom: { enabled },
        zoomX: { enabled }
      };
    } else if ((0, import_is_slot.isSlot)(child, import_Pan.Pan)) {
      const enabled = !disabledOnlyWhenSpecified(child.props);
      slotInteractions = { ...slotInteractions, pan: { enabled } };
    } else if ((0, import_is_slot.isSlot)(child, import_Legend.Legend)) {
      const legend = child.props;
      slots = {
        ...slots,
        legend
      };
    } else if ((0, import_is_slot.isSlot)(child, import_Intent.Intent)) {
      intents = intents ? [...intents, child.props] : [child.props];
    } else if ((0, import_is_slot.isSlot)(child, import_XAxis.XAxis)) {
      const xAxis = child.props;
      slots = {
        ...slots,
        xAxis
      };
    } else if ((0, import_is_slot.isSlot)(child, import_YAxis.YAxis)) {
      const yAxis = child.props;
      slots.yAxis.push(yAxis);
    } else if ((0, import_is_slot.isSlot)(child, import_Tooltip2.Tooltip)) {
      const tooltip = child.props;
      slots = {
        ...slots,
        tooltip
      };
    } else if ((0, import_is_slot.isSlot)(child, import_Annotations.Annotations)) {
      const { children: children2, ...annotationsProps } = child.props;
      const { tracks, tooltip } = iterateTrackSlots(children2);
      const annotations = {
        ...annotationsProps,
        tracks,
        tooltip
      };
      slots = {
        ...slots,
        annotations
      };
    } else if ((0, import_is_slot.isSlot)(child, import_ThresholdIndicator.ThresholdIndicator)) {
      const thresholdProps = child.props;
      slots.thresholds.push(thresholdProps);
    } else if ((0, import_is_slot.isSlot)(child, import_Select.Select)) {
      const select = child.props;
      slots = {
        ...slots,
        select
      };
    } else if ((0, import_is_slot.isSlot)(child, import_ColorRule.ColorRule)) {
      const colorRule = child.props;
      if ((0, import_ColorRule.isValidColorRule)(colorRule)) {
        slots.colorRules.push(colorRule);
      }
    }
  });
  const hasInteractionConfig = Object.keys(chartInteractions).length > 0 || Object.keys(slotInteractions).length > 0;
  slots.interactions = hasInteractionConfig ? getInteractionsConfig(chartInteractions, slotInteractions) : {};
  const hasToolbarConfig = !(0, import_lodash_es.isEmpty)(slotToolbar) || !(0, import_lodash_es.isEmpty)(chartToolbar) || intents;
  if (hasToolbarConfig) {
    slots.toolbar = {
      ...chartToolbar,
      ...slotToolbar,
      enabled: slotToolbar.enabled ?? chartToolbar.enabled ?? true,
      intents
    };
  }
  return slots;
};
const iterateTrackSlots = (children) => {
  const tracks = [];
  let tooltip = {
    template: void 0,
    hidden: false
  };
  import_react.Children.forEach(children, (child) => {
    if ((0, import_is_slot.isSlot)(child, import_Track.Track)) {
      const { children: children2, ...trackProps } = child.props;
      const currentTrack = {
        ...trackProps,
        markers: []
      };
      import_react.Children.forEach(children2, (child2) => {
        if ((0, import_is_slot.isSlot)(child2, import_Marker.Marker)) {
          const marker = (0, import_populate_missing_marker_props.populateMissingMarkerProps)(child2.props, trackProps);
          currentTrack.markers.push(marker);
        }
      });
      const track = (0, import_populate_missing_track_props.populateMissingTrackProps)(currentTrack);
      tracks.push(track);
    } else if ((0, import_is_slot.isSlot)(child, import_Tooltip.AnnotationsTooltip)) {
      const { children: children2, hidden } = child.props;
      tooltip = {
        template: children2,
        hidden: hidden ?? false
      };
    }
  });
  return { tracks, tooltip };
};
const iterateChartToolbarSlots = (children) => {
  let toolbar = {};
  import_react.Children.forEach(children, (childToolbar) => {
    if ((0, import_react_is.isElement)(childToolbar) && childToolbar.type === import_DownloadData.DownloadData) {
      toolbar = {
        ...toolbar,
        downloadData: { enabled: true }
      };
    }
  });
  return toolbar;
};
const iterateChartSeriesSlots = (children) => {
  const line = [];
  const bar = [];
  const area = [];
  const band = [];
  const shapePriority = [];
  import_react.Children.forEach(children, (childSlot) => {
    if ((0, import_react_is.isElement)(childSlot)) {
      switch (childSlot.type) {
        case import_Line.Line: {
          const { data, ...overrides } = childSlot.props;
          line.push({ ...data, ...overrides });
          if (!shapePriority.includes("line")) {
            shapePriority.push("line");
          }
          break;
        }
        case import_Bar.Bar: {
          const { data, ...overrides } = childSlot.props;
          bar.push({ ...data, ...overrides });
          if (!shapePriority.includes("bar")) {
            shapePriority.push("bar");
          }
          break;
        }
        case import_Area.Area: {
          const { data, ...overrides } = childSlot.props;
          area.push({ ...data, ...overrides });
          if (!shapePriority.includes("area")) {
            shapePriority.push("area");
          }
          break;
        }
        case import_Band.Band: {
          const {
            data,
            strokeOnly = false,
            color,
            gapPolicy
          } = childSlot.props;
          band.push({ ...data, strokeOnly, color, gapPolicy });
          if (!shapePriority.includes("band")) {
            shapePriority.push("band");
          }
          break;
        }
      }
    }
  });
  return { band, line, area, bar, shapePriority };
};
