// @dynatrace/strato-components/charts/timeseries/utils/timeseries.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var timeseries_exports = {};
__export(timeseries_exports, {
  getDatapointTime: () => getDatapointTime,
  hasMultiGeometries: () => hasMultiGeometries,
  hasSeriesWithDatapoints: () => hasSeriesWithDatapoints,
  isEmptyTimeseries: () => isEmptyTimeseries,
  mergeTimeseriesFromVariants: () => mergeTimeseriesFromVariants,
  splitTimeseriesPerVariant: () => splitTimeseriesPerVariant
});
module.exports = __toCommonJS(timeseries_exports);
var import_iterate_config_slots = require("./iterate-config-slots.js");
var import_constants = require("../constants.js");
function isEmptyTimeseries(timeseries) {
  return timeseries.length === 0 || timeseries.every((series) => series.datapoints.length === 0);
}
function splitTimeseriesPerVariant(timeseries, children, defaultVariant) {
  const seriesFromSlots = (0, import_iterate_config_slots.iterateChartSeriesSlots)(children);
  const timeseriesPerVariant = {
    line: [],
    area: [],
    bar: [],
    band: [],
    shapePriority: []
  };
  timeseriesPerVariant[defaultVariant ?? import_constants.TIMESERIES_CHART_DEFAULT_VARIANT] = timeseries;
  const shapePriority = defaultVariant && seriesFromSlots.shapePriority.includes(defaultVariant) ? seriesFromSlots.shapePriority : [...seriesFromSlots.shapePriority, defaultVariant];
  return {
    line: timeseriesPerVariant.line.concat(seriesFromSlots.line),
    area: timeseriesPerVariant.area.concat(seriesFromSlots.area),
    bar: timeseriesPerVariant.bar.concat(seriesFromSlots.bar),
    band: timeseriesPerVariant.band.concat(seriesFromSlots.band),
    shapePriority
  };
}
function getDatapointTime(start, end, center) {
  if (center) {
    return center.getTime();
  }
  if (end) {
    return start.getTime() + (end.getTime() - start.getTime()) / 2;
  }
  return start.getTime();
}
const hasSeriesWithDatapoints = (dataPerVariant) => {
  const data = mergeTimeseriesFromVariants(dataPerVariant);
  return data.length > 0 && data.some((series) => series.datapoints.length);
};
function mergeTimeseriesFromVariants(dataPerVariant) {
  const { line, area, bar, band } = dataPerVariant;
  return line.concat(area, bar, band);
}
function hasMultiGeometries(data) {
  const filteredData = Object.keys(data).map(
    (variant) => variant !== "shapePriority" ? data[variant] : []
  );
  const valuesPerVariant = filteredData.filter(
    (geometryData) => geometryData.length > 0
  );
  return valuesPerVariant.length > 1;
}
