// @dynatrace/strato-components/charts/timeseries/utils/yScaleDomain.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var yScaleDomain_exports = {};
__export(yScaleDomain_exports, {
  getBandMinMax: () => getBandMinMax,
  getLineMinMax: () => getLineMinMax,
  getStackedMinMax: () => getStackedMinMax,
  getYAxisValueDomainsByUnit: () => getYAxisValueDomainsByUnit
});
module.exports = __toCommonJS(yScaleDomain_exports);
var import_d3_array = require("d3-array");
var import_safe_get = require("../../core/utils/safe-get.js");
const getValueOrGap = (value, isGap) => isGap ? void 0 : value ?? 0;
const getYTimeSeries = (datapoint, datapointsMap) => {
  const metadata = (0, import_safe_get.safeGet)(datapointsMap, datapoint);
  return getValueOrGap(metadata.value, metadata.isGap);
};
const getYBand0 = (datapoint, datapointsMap) => {
  const metadata = (0, import_safe_get.safeGet)(datapointsMap, datapoint);
  return getValueOrGap(metadata.y0, metadata.isGap);
};
const getYBand1 = (datapoint, datapointsMap) => {
  const metadata = (0, import_safe_get.safeGet)(datapointsMap, datapoint);
  return getValueOrGap(metadata.y1, metadata.isGap);
};
const getDomainByUnit = (unit, seriesMap, datapointsMap, data) => {
  const datapoints = data.series.filter((singleSeries) => (0, import_safe_get.safeGet)(seriesMap, singleSeries).unit === unit).flatMap(
    (currentSeries) => currentSeries.datapoints
  );
  const { min, max } = getMinMax(datapointsMap, {
    variant: data.variant,
    datapoints
  });
  return {
    min,
    max,
    unit
  };
};
function getYAxisValueDomainsByUnit(series, variant, leftAxisMetadata, rightAxisMetadata, seriesMap, datapointsMap) {
  if (series.length === 0) {
    return [];
  }
  const otherGeometriesUnits = [
    leftAxisMetadata.unit,
    rightAxisMetadata.unit
  ].filter((u) => u !== void 0);
  const seriesUnit = series.map((_series) => (0, import_safe_get.safeGet)(seriesMap, _series).unit);
  const scaleUnits = [.../* @__PURE__ */ new Set([...otherGeometriesUnits, ...seriesUnit])];
  return scaleUnits.map(
    (unit) => getDomainByUnit(unit, seriesMap, datapointsMap, {
      series,
      variant
    })
  );
}
const getLineMinMax = (datapointsMap, datapoints) => {
  const [min = 0, max = 0] = (0, import_d3_array.extent)(
    datapoints,
    (dp) => getYTimeSeries(dp, datapointsMap)
  );
  return { min, max };
};
const getStackedMinMax = (datapointsMap, datapoints) => {
  const [min = 0, max = 0] = (0, import_d3_array.extent)(
    datapoints.map((datapoint) => {
      const metadataDatapoint = (0, import_safe_get.safeGet)(datapointsMap, datapoint);
      if (metadataDatapoint.isGap) {
        return void 0;
      }
      return metadataDatapoint.cumulativeValue ?? 0;
    })
  );
  return { min, max };
};
const getBandMinMax = (datapointsMap, datapoints) => {
  const [minY0 = 0, maxY0 = 0] = (0, import_d3_array.extent)(
    datapoints,
    (dp) => getYBand0(dp, datapointsMap)
  );
  const [minY1 = 0, maxY1 = 0] = (0, import_d3_array.extent)(
    datapoints,
    (dp) => getYBand1(dp, datapointsMap)
  );
  const min = Math.min(minY0, minY1);
  const max = Math.max(maxY0, maxY1);
  return { min, max };
};
const MinMaxStrategyMap = {
  band: getBandMinMax,
  line: getLineMinMax,
  area: getStackedMinMax,
  bar: getStackedMinMax
};
const getMinMax = (datapointsMap, data) => {
  return MinMaxStrategyMap[data.variant](
    datapointsMap,
    data.datapoints
  );
};
