// @dynatrace/strato-components/charts/xy-chart/types/axis.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var axis_exports = {};
__export(axis_exports, {
  CATEGORICAL: () => CATEGORICAL,
  LINEAR: () => LINEAR,
  LOG: () => LOG,
  NUMERICAL: () => NUMERICAL,
  TIME: () => TIME,
  XYAxisMapper: () => XYAxisMapper,
  XYAxisTypeCategorical: () => XYAxisTypeCategorical,
  XYAxisTypeLinear: () => XYAxisTypeLinear,
  XYAxisTypeLog: () => XYAxisTypeLog,
  XYAxisTypeTime: () => XYAxisTypeTime
});
module.exports = __toCommonJS(axis_exports);
var import_xy_chart = require("./xy-chart.js");
const NUMERICAL = "numerical";
const LINEAR = "linear";
const CATEGORICAL = "categorical";
const TIME = "time";
const LOG = "log";
const XYAxisTypeTime = TIME;
const XYAxisTypeCategorical = CATEGORICAL;
const XYAxisTypeLinear = `${NUMERICAL}.${LINEAR}`;
const XYAxisTypeLog = `${NUMERICAL}.${LOG}`;
const XYAxisMapper = {
  [import_xy_chart.NUMERICAL_SCALE_TYPE]: XYAxisTypeLinear,
  [import_xy_chart.LOG_SCALE_TYPE]: XYAxisTypeLog,
  [import_xy_chart.CATEGORICAL_SCALE_TYPE]: XYAxisTypeCategorical,
  [import_xy_chart.TIME_SCALE_TYPE]: XYAxisTypeTime
};
