// @dynatrace/strato-components/charts/xy-chart/types/guards.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var guards_exports = {};
__export(guards_exports, {
  isAnyNumber: () => isAnyNumber,
  isAnyNumericalAxis: () => isAnyNumericalAxis,
  isAnyNumericalScale: () => isAnyNumericalScale,
  isAreaContentPoint: () => isAreaContentPoint,
  isCategorical: () => isCategorical,
  isCategoricalAxis: () => isCategoricalAxis,
  isCategoricalScale: () => isCategoricalScale,
  isDate: () => isDate,
  isDotContentPoint: () => isDotContentPoint,
  isDotOrLineSelection: () => isDotOrLineSelection,
  isFunction: () => isFunction,
  isLineContentPoint: () => isLineContentPoint,
  isLinear: () => isLinear,
  isLinearAxis: () => isLinearAxis,
  isLogarithmic: () => isLogarithmic,
  isLogarithmicAxis: () => isLogarithmicAxis,
  isNumber: () => isNumber,
  isNumericColoringScale: () => isNumericColoringScale,
  isNumerical: () => isNumerical,
  isNumericalAxis: () => isNumericalAxis,
  isString: () => isString,
  isTime: () => isTime,
  isTimeAxis: () => isTimeAxis,
  isTimeScale: () => isTimeScale,
  isValidCategoryValue: () => isValidCategoryValue
});
module.exports = __toCommonJS(guards_exports);
var import_xy_chart_internals = require("./xy-chart-internals.js");
const isNumber = (value) => {
  return typeof value === "number";
};
const isString = (value) => {
  return typeof value === "string";
};
const isDate = (value) => {
  return value instanceof Date;
};
const isFunction = (fn) => {
  return typeof fn === "function";
};
const REGEX_LINEAR = /linear/;
const REGEX_LOG = /log/;
const REGEX_TIME = /^time$/;
const REGEX_CATEGORICAL = /^categorical$/;
const REGEX_NUMERICAL = /numerical/;
const REGEX_ANY_NUMBER = /^numerical|^time$/;
const isLinear = (axisType) => REGEX_LINEAR.test(axisType);
const isLogarithmic = (axisType) => REGEX_LOG.test(axisType);
const isTime = (axisType) => REGEX_TIME.test(axisType);
const isCategorical = (axisType) => REGEX_CATEGORICAL.test(axisType);
const isNumerical = (axisType) => REGEX_NUMERICAL.test(axisType);
const isAnyNumber = (axisType) => REGEX_ANY_NUMBER.test(axisType);
function isLinearAxis(axis) {
  return isLinear(axis.type);
}
function isLogarithmicAxis(axis) {
  return isLogarithmic(axis.type);
}
function isNumericalAxis(axis) {
  return isNumerical(axis.type);
}
function isAnyNumericalAxis(axis) {
  return !isCategorical(axis.type);
}
function isTimeAxis(axis) {
  return isTime(axis.type);
}
function isCategoricalAxis(axis) {
  return isCategorical(axis.type);
}
function isCategoricalScale(args) {
  return !!args.axisType && isCategoricalAxis({ type: args.axisType });
}
function isTimeScale(args) {
  return !!args.axisType && isTime(args.axisType);
}
function isAnyNumericalScale(args) {
  return !!args.axisType && isAnyNumericalAxis({ type: args.axisType });
}
function isDotContentPoint(content) {
  return !!content && content.type === import_xy_chart_internals.DOT_VARIANT && !!content.series && !!content.datapoint;
}
function isLineContentPoint(content) {
  return !!content && content.type === import_xy_chart_internals.LINE_VARIANT && !!content.series && !!content.datapoint;
}
function isAreaContentPoint(content) {
  return !!content && content.type === import_xy_chart_internals.AREA_VARIANT && !!content.series && !!content.datapoint;
}
function isNumericColoringScale(colorScale) {
  return colorScale.domain().every((value) => typeof value === "number");
}
const isValidCategoryValue = (value) => Boolean(value && typeof value === "string");
function isDotOrLineSelection(point) {
  return point.series.type === import_xy_chart_internals.DOT_VARIANT || point.series.type === import_xy_chart_internals.LINE_VARIANT;
}
