// @dynatrace/strato-components/charts/xy-chart/types/xy-chart.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var xy_chart_exports = {};
__export(xy_chart_exports, {
  BAR_VARIANT_GROUP: () => BAR_VARIANT_GROUP,
  BAR_VARIANT_SINGLE: () => BAR_VARIANT_SINGLE,
  BAR_VARIANT_STACK: () => BAR_VARIANT_STACK,
  BOTH_POSITION: () => BOTH_POSITION,
  BOTTOM_POSITION: () => BOTTOM_POSITION,
  CATEGORICAL_SCALE_TYPE: () => CATEGORICAL_SCALE_TYPE,
  LEFT_POSITION: () => LEFT_POSITION,
  LOG_SCALE_TYPE: () => LOG_SCALE_TYPE,
  NUMERICAL_SCALE_TYPE: () => NUMERICAL_SCALE_TYPE,
  RIGHT_POSITION: () => RIGHT_POSITION,
  TIME_SCALE_TYPE: () => TIME_SCALE_TYPE,
  TOP_POSITION: () => TOP_POSITION
});
module.exports = __toCommonJS(xy_chart_exports);
const BAR_VARIANT_SINGLE = "single";
const BAR_VARIANT_STACK = "stack";
const BAR_VARIANT_GROUP = "group";
const NUMERICAL_SCALE_TYPE = "numerical";
const LOG_SCALE_TYPE = "log";
const TIME_SCALE_TYPE = "time";
const CATEGORICAL_SCALE_TYPE = "categorical";
const TOP_POSITION = "top";
const BOTTOM_POSITION = "bottom";
const LEFT_POSITION = "left";
const RIGHT_POSITION = "right";
const BOTH_POSITION = "both";
