// @dynatrace/strato-components/charts/xy-chart/types/xy-chart-internals.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var xy_chart_internals_exports = {};
__export(xy_chart_internals_exports, {
  AREA_HORIZONTAL: () => AREA_HORIZONTAL,
  AREA_ID: () => AREA_ID,
  AREA_VARIANT: () => AREA_VARIANT,
  AREA_VERTICAL: () => AREA_VERTICAL,
  BAND_HORIZONTAL: () => BAND_HORIZONTAL,
  BAND_VARIANT: () => BAND_VARIANT,
  BAND_VERTICAL: () => BAND_VERTICAL,
  BAR_ID: () => BAR_ID,
  BAR_VARIANT: () => BAR_VARIANT,
  CATEGORICAL_LEGEND_TYPE: () => CATEGORICAL_LEGEND_TYPE,
  DOT_ID: () => DOT_ID,
  DOT_VARIANT: () => DOT_VARIANT,
  LINE_ID: () => LINE_ID,
  LINE_VARIANT: () => LINE_VARIANT,
  RANGE_LEGEND_TYPE: () => RANGE_LEGEND_TYPE,
  RECT_ID: () => RECT_ID,
  RECT_VARIANT: () => RECT_VARIANT
});
module.exports = __toCommonJS(xy_chart_internals_exports);
var import_canvas_shape = require("../../core/components/canvas/types/canvas-shape.js");
const RANGE_LEGEND_TYPE = "range";
const CATEGORICAL_LEGEND_TYPE = "categorical";
const RECT_VARIANT = "rect";
const DOT_VARIANT = "dot";
const LINE_VARIANT = "line";
const BAR_VARIANT = "bar";
const AREA_VARIANT = "area";
const BAND_VARIANT = "band";
const AREA_HORIZONTAL = `${AREA_VARIANT}${import_canvas_shape.ORIENTATION_HORIZONTAL}`;
const AREA_VERTICAL = `${AREA_VARIANT}${import_canvas_shape.ORIENTATION_VERTICAL}`;
const BAND_HORIZONTAL = `${BAND_VARIANT}${import_canvas_shape.ORIENTATION_HORIZONTAL}`;
const BAND_VERTICAL = `${BAND_VARIANT}${import_canvas_shape.ORIENTATION_VERTICAL}`;
const DOT_ID = "dot_id";
const LINE_ID = "line_id";
const AREA_ID = "area_id";
const RECT_ID = "rect_id";
const BAR_ID = "bar_id";
