// @dynatrace/strato-components/esm/charts/core/components/annotations/components/AnnotationsRenderer.js
import { jsx } from "react/jsx-runtime";
import { useOverlay } from "../../overlay-tooltip/contexts/OverlayContext.js";
import { AnnotationsAtom } from "../AnnotationsAtom.js";
import { useAnnotationsSelectionState } from "../hooks/store/selectors.js";
import { useGroupedAnnotationsData } from "../hooks/useGroupedAnnotationsData.js";
import { useMarkerEventHandlers } from "../hooks/useMarkerEventHandlers.js";
import { MarkerAtom } from "../slots/MarkerAtom.js";
import { TrackAtom } from "../slots/TrackAtom.js";
import { calculateMarkerDimension } from "../utils/calculate-marker-dimension.js";
import { defineMarkerBrightnessState } from "../utils/define-marker-brightness-state.js";
import { defineMarkerColor } from "../utils/define-marker-color.js";
import { defineMarkerContent } from "../utils/define-marker-content.js";
const AnnotationsRenderer = (props) => {
  const {
    scale,
    width,
    labelsOffset,
    showLabels,
    tracks,
    textOverflow,
    truncateMode
  } = props;
  const overlay = useOverlay();
  const groupedData = useGroupedAnnotationsData(tracks, scale.domain());
  const selection = useAnnotationsSelectionState();
  const { handleMouseEnter, handleMouseMove, handleMouseLeave, handleClick } = useMarkerEventHandlers(overlay.overlayRef);
  return /* @__PURE__ */ jsx(
    AnnotationsAtom,
    {
      width,
      labelsWidth: labelsOffset,
      showLabels,
      children: groupedData.map((track, trackNumber) => {
        const currentTrack = Array.from(track.markers.entries());
        return /* @__PURE__ */ jsx(
          TrackAtom,
          {
            label: track.label,
            hidden: track.hidden,
            children: currentTrack.map(([groupId, annotationsGroup]) => {
              return [...annotationsGroup].sort((a, b) => a.priority - b.priority).map(({ key: markerId, markers }) => {
                const [marker] = markers;
                const color = defineMarkerColor(markers);
                const state = defineMarkerBrightnessState(
                  markerId,
                  selection
                );
                const { x, width: width2 } = calculateMarkerDimension(marker, scale);
                const markerContent = defineMarkerContent(markers);
                return /* @__PURE__ */ jsx(
                  MarkerAtom,
                  {
                    x,
                    width: width2,
                    color,
                    onMouseEnter: handleMouseEnter(
                      markerId,
                      groupId,
                      annotationsGroup
                    ),
                    onMouseMove: handleMouseMove,
                    onMouseLeave: handleMouseLeave,
                    onClick: handleClick(groupId, annotationsGroup),
                    content: markerContent,
                    interactionState: state,
                    textOverflow,
                    truncateMode
                  },
                  markerId
                );
              });
            })
          },
          trackNumber
        );
      })
    }
  );
};
AnnotationsRenderer.displayName = "AnnotationsContent";
export {
  AnnotationsRenderer
};
