// @dynatrace/strato-components/charts/core/components/legend/categorical/ChartLegend.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ChartLegend_exports = {};
__export(ChartLegend_exports, {
  ChartLegend: () => ChartLegend
});
module.exports = __toCommonJS(ChartLegend_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_ChartLegend_css = require("./ChartLegend.sty.js");
var import_ChartLegendGrid = require("./components/ChartLegendGrid.js");
var import_ChartVirtualizedLegendGrid = require("./components/ChartVirtualizedLegendGrid.js");
var import_LegendGridSettingsContext = require("./context/LegendGridSettingsContext.js");
var import_useLegendInteractions = require("./hooks/useLegendInteractions.js");
var import_legend_grid_utils = require("./utils/legend-grid.utils.js");
var import_constants = require("../../../constants.js");
var import_Legend_context = require("../shared/context/Legend.context.js");
var import_LegendSizeContext = require("../shared/context/LegendSizeContext.js");
const ChartLegend = (props) => {
  const {
    items,
    orientation = import_constants.ORIENTATION.Horizontal,
    onHover,
    onClick,
    onSeriesAction,
    interactionsDisabled = false,
    truncationMode = import_constants.DEFAULT_TRUNCATION_MODE,
    style
  } = props;
  const ref = (0, import_react.useRef)(null);
  const { width, height } = (0, import_LegendSizeContext.useLegendSize)();
  const { onItemHover, onLegendIconClick, onLegendNameClick, hiddenItems } = (0, import_useLegendInteractions.useLegendInteractions)(items, { onClick, onHover });
  const gridSettings = (0, import_legend_grid_utils.buildGridSettings)(items, width, height);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      ref,
      "data-testid": "chart-legend",
      className: import_ChartLegend_css.legendContainer,
      style: {
        height,
        width,
        flexDirection: orientation === import_constants.ORIENTATION.Vertical ? "row" : "column",
        ...style
      },
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_Legend_context.LegendContext.Provider,
        {
          value: {
            hiddenItems,
            truncationMode,
            onSeriesAction,
            onLegendIconClick,
            onItemHover,
            onLegendNameClick,
            interactionsDisabled
          },
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_LegendGridSettingsContext.LegendGridSettingsContext.Provider, { value: gridSettings, children: gridSettings.hasVirtual ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_ChartVirtualizedLegendGrid.ChartVirtualizedLegendGrid,
            {
              items,
              orientation
            }
          ) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ChartLegendGrid.ChartLegendGrid, { items, orientation }) })
        }
      )
    }
  );
};
ChartLegend["displayName"] = "ChartLegend";
