// @dynatrace/strato-components/charts/core/components/legend/categorical/constants.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var constants_exports = {};
__export(constants_exports, {
  CATEGORICAL_LEGEND_CLASS: () => CATEGORICAL_LEGEND_CLASS,
  ITEM_HEIGHT: () => ITEM_HEIGHT,
  LABEL_ITEM_CLASSNAME: () => LABEL_ITEM_CLASSNAME,
  LEGEND_ACTION_BUTTON_WIDTH: () => LEGEND_ACTION_BUTTON_WIDTH,
  LEGEND_BUTTON_WIDTH: () => LEGEND_BUTTON_WIDTH,
  LEGEND_ICON_WIDTH: () => LEGEND_ICON_WIDTH,
  MIN_ITEM_WIDTH: () => MIN_ITEM_WIDTH,
  VIRTUALIZED_GRID_THRESHOLD: () => VIRTUALIZED_GRID_THRESHOLD
});
module.exports = __toCommonJS(constants_exports);
var import_spacings = __toESM(require("@dynatrace/strato-design-tokens/spacings"));
const ITEM_HEIGHT = { value: 20, token: import_spacings.default.Size20 };
const MIN_ITEM_WIDTH = 120;
const VIRTUALIZED_GRID_THRESHOLD = 50;
const LEGEND_BUTTON_WIDTH = 50;
const LEGEND_ICON_WIDTH = 25;
const LEGEND_ACTION_BUTTON_WIDTH = 16;
const LABEL_ITEM_CLASSNAME = "chart-legend-item-label";
const CATEGORICAL_LEGEND_CLASS = "categorical-legend";
