// @dynatrace/strato-components/charts/core/components/legend/categorical/messages.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var messages_exports = {};
__export(messages_exports, {
  LEGEND_MESSAGES: () => LEGEND_MESSAGES
});
module.exports = __toCommonJS(messages_exports);
var import_react_intl = require("react-intl");
const LEGEND_MESSAGES = (0, import_react_intl.defineMessages)({
  legendHideSeries: {
    id: "XPLMI3aOe8TwFwJc",
    defaultMessage: "Hide",
    description: "Text in a tooltip shown when hovering a legend series to hide it"
  },
  legendShowSeries: {
    id: "1e3uPo8SoI4NA2dZ",
    defaultMessage: "Show",
    description: "Text in a tooltip shown when hovering a legend series to show it"
  }
});
