// @dynatrace/strato-components/charts/single-value-grid/utils/build-props-from-grid.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var build_props_from_grid_exports = {};
__export(build_props_from_grid_exports, {
  buildPropsFromGrid: () => buildPropsFromGrid
});
module.exports = __toCommonJS(build_props_from_grid_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_Menu = require("../../../navigation/menu/Menu.js");
var import_ToolbarMenu_css = require("../../core/components/toolbar/components/ToolbarMenu.sty.js");
var import_render_menu_action_children = require("../../core/utils/render-menu-action-children.js");
const getIntentsAsyncTemplate = (singleIntents, gridIntents, data, labelText) => async () => {
  const intentsTemplate = gridIntents.length ? [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Group, { className: import_ToolbarMenu_css.toolbarMenuGroup, children: gridIntents.map((item, key) => /* @__PURE__ */ (0, import_react.createElement)(import_Menu.Menu.Intent, { ...item, key })) }, "group")
  ] : [];
  if (singleIntents) {
    return (async () => singleIntents(data))().then((template) => {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, { children: template ? [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_Menu.Menu.Group,
          {
            className: import_ToolbarMenu_css.toolbarMenuGroup,
            children: (0, import_render_menu_action_children.renderMenuActionChildren)(template.props.children, {
              divideActionsAndIntents: false
            })
          },
          "individualIntents"
        ),
        gridIntents.length ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Label, { children: labelText }, "label") : null,
        ...intentsTemplate
      ] : intentsTemplate });
    }).catch(() => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, { children: intentsTemplate }));
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, { children: intentsTemplate });
};
const buildPropsFromGrid = (toolbar, singleIntents, gridIntents, labelText) => (data) => {
  if (!singleIntents) {
    return void 0;
  }
  return {
    intentsAsyncTemplate: getIntentsAsyncTemplate(
      singleIntents,
      gridIntents,
      data,
      labelText
    ),
    toolbar: {
      ...toolbar,
      draggable: false
    }
  };
};
