// @dynatrace/strato-components/charts/single-value-grid/utils/iterate-slots.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var iterate_slots_exports = {};
__export(iterate_slots_exports, {
  iterateSlots: () => iterateSlots
});
module.exports = __toCommonJS(iterate_slots_exports);
var import_react = require("react");
var import_Intent = require("../../../core/slots/Intent/Intent.js");
var import_is_slot = require("../../../core/utils/is-slot.js");
var import_iterate_sparkline_slots = require("../../core/components/sparkline-display/utils/iterate-sparkline-slots.js");
var import_ColorRule = require("../../core/slots/ColorRule.js");
var import_CompactToolbar = require("../../core/slots/toolbar/CompactToolbar.js");
var import_extract_behavior_tracking_props = require("../../core/utils/extract-behavior-tracking-props.js");
var import_Sparkline = require("../slots/Sparkline.js");
var import_Toolbar = require("../slots/Toolbar.js");
var import_Trend = require("../slots/Trend.js");
var import_Value = require("../slots/Value.js");
const iterateSlots = (children, props) => {
  const slots = {
    intents: { grid: [], single: void 0, none: true },
    toolbar: {
      hidden: false,
      placement: (props.alignment || "") === "end" ? "top-left" : "top-right",
      "data-dt-content": "",
      "data-dt-name": "",
      "data-dt-children-name": "",
      "data-dt-features": "",
      "data-dt-properties": "",
      "data-dt-mouse-over": "",
      "data-dt-value-content": false,
      "data-dt-children-value-content": false,
      "data-dt-ancestor": ""
    }
  };
  const accessors = {};
  const colorRules = [];
  import_react.Children.forEach(children, (child) => {
    if ((0, import_is_slot.isSlot)(child, import_Value.Value)) {
      const { dataAccessor, formatterAccessor, unitAccessor, ...valueProps } = child.props;
      slots.value = valueProps;
      accessors.value = {
        dataAccessor,
        formatterAccessor,
        unitAccessor
      };
    } else if ((0, import_is_slot.isSlot)(child, import_Sparkline.Sparkline)) {
      const {
        children: sparklineSlots,
        dataAccessor,
        colorAccessor,
        ...sparklineProps
      } = child.props;
      const { xAxis, yAxis } = (0, import_iterate_sparkline_slots.iterateSparklineSlots)(sparklineSlots);
      slots.sparkline = sparklineProps;
      accessors.sparkline = {
        dataAccessor,
        colorAccessor,
        xAxis,
        yAxis
      };
    } else if ((0, import_is_slot.isSlot)(child, import_Trend.Trend)) {
      const {
        directionAccessor,
        valueAccessor,
        labelAccessor,
        formatterAccessor,
        inverseTrendAccessor,
        ...trendProps
      } = child.props;
      slots.trend = trendProps;
      accessors.trend = {
        directionAccessor,
        valueAccessor,
        labelAccessor,
        formatterAccessor,
        inverseTrendAccessor
      };
    } else if ((0, import_is_slot.isSlot)(child, import_ColorRule.ColorRule)) {
      const colorRule = child.props;
      if ((0, import_ColorRule.isValidColorRule)(colorRule)) {
        colorRules.push(colorRule);
      }
    } else if ((0, import_is_slot.isSlot)(child, import_Intent.Intent)) {
      slots.intents.grid.push(child.props);
      slots.intents.none = false;
    } else if ((0, import_is_slot.isSlot)(child, import_Toolbar.SingleValueGridToolbarSlot)) {
      slots.toolbar = {
        ...slots.toolbar,
        ...(0, import_CompactToolbar.sanitizeProps)(child.props),
        ...(0, import_extract_behavior_tracking_props.extractBehaviorTrackingProps)(child.props)
      };
    }
  });
  if (colorRules.length > 0) {
    slots.colorRules = colorRules;
  }
  if (slots.toolbar.hidden) {
    slots.intents = {
      ...slots.intents,
      grid: [],
      none: true
    };
  } else {
    if (props.seriesActions) {
      slots.intents = {
        ...slots.intents,
        single: props.seriesActions,
        none: false
      };
    }
    if (slots.intents.none) {
      slots.toolbar = {
        ...slots.toolbar,
        hidden: true
      };
    }
  }
  return { ...slots, accessors };
};
