// @dynatrace/strato-components/charts/core/components/annotations/components/canvas/displays/AnnotationsMarkersDisplay.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var AnnotationsMarkersDisplay_exports = {};
__export(AnnotationsMarkersDisplay_exports, {
  AnnotationsMarkersDisplay: () => AnnotationsMarkersDisplay
});
module.exports = __toCommonJS(AnnotationsMarkersDisplay_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_useCurrentTheme = require("../../../../../../../core/hooks/useCurrentTheme.js");
var import_getTokenComputedStyle = require("../../../../../utils/getTokenComputedStyle.js");
var import_RectClipPathCanvas = require("../../../../canvas/components/RectClipPathCanvas.js");
var import_constants = require("../../../constants.js");
var import_AnnotationsConfig_context = require("../../../contexts/AnnotationsConfig.context.js");
var import_crop_marker = require("../../../utils/crop-marker.js");
var import_get_markers_offset_by_track = require("../../../utils/get-markers-offset-by-track.js");
var import_use_track_display = require("../hooks/use-track-display.js");
var import_CanvasContextRestore = require("../renderers/CanvasContextRestore.js");
var import_RangeMarkerRenderer = require("../renderers/RangeMarkerRenderer.js");
var import_constants2 = require("../renderers/value-marker-renderer/constants.js");
var import_ValueMarkerRenderer = require("../renderers/value-marker-renderer/ValueMarkerRenderer.js");
var import_calculate_marker_width_and_content = require("../utils/calculate-marker-width-and-content.js");
var import_get_marker_brightness_percentage_based_on_state = require("../utils/get-marker-brightness-percentage-based-on-state.js");
const MARKER_DEFAULT_OPACITY = 0.85;
const MARKER_HOVER_OPACITY = 0.2;
const AnnotationsMarkersDisplay = ({
  markers
}) => {
  const theme = (0, import_useCurrentTheme.useCurrentTheme)();
  const { trackWidth } = (0, import_react.useContext)(import_AnnotationsConfig_context.AnnotationsConfigContext);
  const { offsetX, trackIndex } = (0, import_use_track_display.useTrackDisplay)();
  return markers.map((props, markerKey) => {
    const {
      color,
      content,
      interactionState = "default",
      width: widthProp,
      x,
      textOverflow,
      truncateMode
    } = props;
    const isRange = widthProp > 0;
    const outOfBounds = isRange ? x + widthProp <= 0 : x < 0 || x > trackWidth;
    if (outOfBounds) {
      return null;
    }
    const markerMinWidth = isRange ? import_constants.MINIMUM_MARKER_WIDTH : import_constants2.VALUE_MARKER_WIDTH;
    const width = Math.max(widthProp, markerMinWidth);
    const markerOffsetByTrack = (0, import_get_markers_offset_by_track.getMarkerOffsetByTrack)(trackIndex);
    const brightness = (0, import_get_marker_brightness_percentage_based_on_state.getMarkerBrightnessPercentageBasedOnState)(
      theme,
      interactionState
    );
    const opacity = interactionState === "inactive" ? MARKER_HOVER_OPACITY : MARKER_DEFAULT_OPACITY;
    const { croppedX, croppedWidth } = (0, import_crop_marker.cropMarker)(x, width, trackWidth);
    const {
      sanitizedContent,
      calculatedMarkerWidth,
      valueMarkerOrientation,
      calculatedX
    } = (0, import_calculate_marker_width_and_content.calculateMarkerWidthAndContent)({
      trackWidth,
      trackLeftBoundary: offsetX,
      content,
      width: croppedWidth,
      x: croppedX,
      isRange,
      textOverflow,
      truncateMode
    });
    const markerProps = {
      x: calculatedX,
      y: markerOffsetByTrack,
      width: calculatedMarkerWidth,
      height: import_constants.PILL_MARKER_HEIGHT,
      fillColor: (0, import_getTokenComputedStyle.getTokenComputedStyle)(color),
      strokeColor: (0, import_getTokenComputedStyle.getTokenComputedStyle)(import_constants.MARKER_STROKE_COLOR),
      opacity,
      brightness,
      children: sanitizedContent
    };
    return isRange ? /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_react.Fragment, { children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_RectClipPathCanvas.RectClipPathCanvas,
        {
          position: {
            x: offsetX - import_constants.MARKER_STROKE_WIDTH * devicePixelRatio,
            y: markerOffsetByTrack - import_constants.MARKER_STROKE_WIDTH * devicePixelRatio
          },
          height: import_constants.TRACK_HEIGHT,
          width: trackWidth + 2 * import_constants.MARKER_STROKE_WIDTH * devicePixelRatio
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_RangeMarkerRenderer.RangeMarkerRenderer, { ...markerProps }),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_CanvasContextRestore.CanvasContextRestore, {})
    ] }, `${markerKey}_${x}`) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react.Fragment, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_ValueMarkerRenderer.ValueMarkerRenderer,
      {
        ...markerProps,
        orientation: valueMarkerOrientation
      }
    ) }, `${markerKey}_${x}`);
  });
};
AnnotationsMarkersDisplay["displayName"] = "AnnotationsMarkersDisplay";
