// @dynatrace/strato-components/charts/core/components/annotations/components/canvas/displays/AnnotationsTrackDisplay.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var AnnotationsTrackDisplay_exports = {};
__export(AnnotationsTrackDisplay_exports, {
  AnnotationsTrackDisplay: () => AnnotationsTrackDisplay
});
module.exports = __toCommonJS(AnnotationsTrackDisplay_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_AnnotationsMarkersDisplay = require("./AnnotationsMarkersDisplay.js");
var import_getTokenComputedStyle = require("../../../../../utils/getTokenComputedStyle.js");
var import_constants = require("../../../constants.js");
var import_AnnotationsConfig_context = require("../../../contexts/AnnotationsConfig.context.js");
var import_get_track_separator_offset = require("../../../utils/get-track-separator-offset.js");
var import_track_display_context = require("../context/track-display.context.js");
var import_AnnotationsLabelRenderer = require("../renderers/AnnotationsLabelRenderer.js");
var import_AnnotationsLineRenderer = require("../renderers/AnnotationsLineRenderer.js");
const getLabelYPosition = (index) => (0, import_get_track_separator_offset.getTrackSeparatorOffset)(index) + import_constants.LABEL_OFFSET;
const AnnotationsTrackDisplay = (props) => {
  const { drawTopLine, label, markers, trackIndex, emphasizeBottomLine } = props;
  const { labelsWidth, trackWidth, showLabels } = (0, import_react.useContext)(
    import_AnnotationsConfig_context.AnnotationsConfigContext
  );
  const trackFinalXPosition = labelsWidth + trackWidth;
  const topLineY = (0, import_get_track_separator_offset.getTrackSeparatorOffset)(trackIndex);
  const bottomLineY = (0, import_get_track_separator_offset.getTrackSeparatorOffset)(trackIndex + 1);
  const bottomLineColor = emphasizeBottomLine ? import_constants.EMPHASIZED_TRACK_SEPARATOR_COLOR : import_constants.TRACK_SEPARATOR_COLOR;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    drawTopLine && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_AnnotationsLineRenderer.AnnotationsLineRenderer,
      {
        source: { x: labelsWidth, y: topLineY },
        destination: { x: trackFinalXPosition, y: topLineY },
        strokeColor: (0, import_getTokenComputedStyle.getTokenComputedStyle)(import_constants.TRACK_SEPARATOR_COLOR)
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_react.Fragment, { children: [
      showLabels && label && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_AnnotationsLabelRenderer.AnnotationsLabelRenderer,
        {
          text: label,
          x: labelsWidth - import_constants.MARKER_LABEL_PADDING,
          y: getLabelYPosition(trackIndex),
          maxWidth: labelsWidth - import_constants.MARKER_LABEL_PADDING,
          ellipsis: true
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_track_display_context.TrackDisplayContext.Provider,
        {
          value: {
            offsetX: labelsWidth,
            trackIndex
          },
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_AnnotationsMarkersDisplay.AnnotationsMarkersDisplay, { markers })
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_AnnotationsLineRenderer.AnnotationsLineRenderer,
        {
          source: { x: labelsWidth, y: bottomLineY },
          destination: {
            x: trackFinalXPosition,
            y: bottomLineY
          },
          strokeColor: (0, import_getTokenComputedStyle.getTokenComputedStyle)(bottomLineColor)
        }
      )
    ] })
  ] });
};
AnnotationsTrackDisplay["displayName"] = "AnnotationsTrackDisplay";
