// @dynatrace/strato-components/esm/charts/timeseries/components/toolbar/DownloadDataButton.js
import { jsx } from "react/jsx-runtime";
import { useRef } from "react";
import { useIntl } from "react-intl";
import { DownloadIcon } from "@dynatrace/strato-icons";
import { ChartToolbarRenderer } from "../../../core/components/toolbar/ChartToolbarRenderer.js";
import { useA11yToolbarTabIndex } from "../../../core/hooks/useA11yToolbarTabIndex.js";
import { useTimeseriesRawData } from "../../hooks/useTimeseriesRawData.js";
import { TIMESERIES_MESSAGES } from "../../messages/tooltips-text.js";
import { buildAndDownloadCsv } from "../../utils/download/build-and-download-csv.js";
const DownloadDataButton = () => {
  const downloadDataButtonRef = useRef(null);
  const tabIndex = useA11yToolbarTabIndex("download");
  const intl = useIntl();
  const rawData = useTimeseriesRawData();
  const handleDownloadDataKeyboard = (event) => {
    if (event.key === "Enter") {
      buildAndDownloadCsv(rawData);
    }
  };
  const handleDownloadDataToolbar = () => {
    buildAndDownloadCsv(rawData);
  };
  return /* @__PURE__ */ jsx(
    ChartToolbarRenderer.Control,
    {
      ref: downloadDataButtonRef,
      prefixIcon: /* @__PURE__ */ jsx(DownloadIcon, {}),
      tooltipText: intl.formatMessage(TIMESERIES_MESSAGES.downloadAsCsv),
      onKeyDown: handleDownloadDataKeyboard,
      onClick: handleDownloadDataToolbar,
      tabIndex,
      divider: "left"
    }
  );
};
DownloadDataButton["displayName"] = "DownloadDataButton";
export {
  DownloadDataButton
};
