// @dynatrace/strato-components/esm/charts/timeseries/components/toolbar/ResetButton.js
import { jsx } from "react/jsx-runtime";
import { useRef } from "react";
import { useIntl } from "react-intl";
import { ResetIcon } from "@dynatrace/strato-icons";
import { useSynchronizedSelection } from "../../../core/components/shared-chart-interactions/hooks/useSynchronizedSelection.js";
import { ChartToolbarRenderer } from "../../../core/components/toolbar/ChartToolbarRenderer.js";
import { useA11yToolbarTabIndex } from "../../../core/hooks/useA11yToolbarTabIndex.js";
import { isSameDomain } from "../../../core/utils/zoom.js";
import { useCurrentDomainState } from "../../hooks/store/selectors.js";
import { useChartInteractions } from "../../hooks/useChartInteractions.js";
import { useXScales } from "../../hooks/useXScale.js";
import { TIMESERIES_MESSAGES } from "../../messages/tooltips-text.js";
import { useResetInteractions } from "../overlays/hooks/useResetInteractions.js";
const ResetButton = () => {
  const { initialScale } = useXScales();
  const intl = useIntl();
  const resetButtonRef = useRef(null);
  const tabIndex = useA11yToolbarTabIndex("reset");
  const currentDomain = useCurrentDomainState();
  const { zoom, zoomX } = useChartInteractions();
  const resetInteractions = useResetInteractions();
  const { resetSelectionState } = useSynchronizedSelection();
  const hasZoomInteractions = zoom?.enabled || zoomX?.enabled;
  const handleClick = () => {
    resetInteractions();
    resetSelectionState?.();
  };
  function handleZoomResetKeyboardMode(event) {
    if (event.code === "Enter") {
      resetInteractions();
    }
  }
  const isResettable = !isSameDomain(currentDomain, initialScale.domain());
  const onlyXZoomIsEnabled = zoomX?.enabled && !zoom?.enabled;
  return hasZoomInteractions ? /* @__PURE__ */ jsx(
    ChartToolbarRenderer.Control,
    {
      ref: resetButtonRef,
      prefixIcon: /* @__PURE__ */ jsx(ResetIcon, {}),
      tooltipText: intl.formatMessage(TIMESERIES_MESSAGES.zoomActionReset),
      tooltipKeyShortcut: "R",
      testId: "chart-toolbar-zoom-reset",
      onKeyDown: handleZoomResetKeyboardMode,
      onClick: handleClick,
      tabIndex,
      disabled: !isResettable,
      placement: "toolbar",
      divider: onlyXZoomIsEnabled ? "left" : void 0,
      showWhenCollapsed: isResettable
    }
  ) : null;
};
ResetButton["displayName"] = "ResetButton";
export {
  ResetButton
};
