// @dynatrace/strato-components/esm/charts/timeseries/components/toolbar/TimeseriesChartToolbar.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { isUndefined } from "lodash-es";
import { useIntl } from "react-intl";
import { DownloadDataButton } from "./DownloadDataButton.js";
import { ExploreButton } from "./ExploreButton.js";
import { PanButton } from "./PanButton.js";
import { ResetButton } from "./ResetButton.js";
import { ZoomInOutButtons } from "./ZoomInOutButtons.js";
import { ZoomModeButton } from "./ZoomModeButton.js";
import { useBehavioralTrackingProps as _useBehavioralTrackingProps } from "../../../../core/hooks/useBehavioralTrackingProps.js";
import { ChartToolbarRenderer } from "../../../core/components/toolbar/ChartToolbarRenderer.js";
import { getMenuTriggerDisplayMode } from "../../../core/components/toolbar/utils/get-menu-trigger-display-mode.js";
import { getMenuTriggerIcon } from "../../../core/components/toolbar/utils/get-menu-trigger-icon.js";
import { ChartOverlaysElevation } from "../../../core/constants.js";
import { getA11yToolbarConfig } from "../../../core/utils/zoom.js";
import { useInteractionModeState } from "../../hooks/store/selectors.js";
import { useChartInteractions } from "../../hooks/useChartInteractions.js";
import { useChartToolbar } from "../../hooks/useChartToolbar.js";
import { useTimeseriesRawData } from "../../hooks/useTimeseriesRawData.js";
import { useTimeseriesSelectionAreaRef } from "../../hooks/useTimeseriesSelectionAreaRef.js";
import { TIMESERIES_MESSAGES } from "../../messages/tooltips-text.js";
import { buildAndDownloadCsv } from "../../utils/download/build-and-download-csv.js";
const defaultToolbarPlacement = "top-right";
const TimeseriesChartToolbar = ({
  onMouseEnter,
  onMouseLeave,
  handlerRef
}) => {
  const selectionAreaRef = useTimeseriesSelectionAreaRef();
  const config = useChartToolbar();
  const { zoom, zoomX, pan } = useChartInteractions();
  const intl = useIntl();
  const isDownloadEnabled = config?.downloadData?.enabled ?? false;
  const isPanEnabled = pan?.enabled;
  const isZoomEnabled = zoom?.enabled;
  const isZoomXEnabled = zoomX?.enabled;
  const anyInteractionEnabled = isZoomEnabled || isZoomXEnabled || isPanEnabled || isDownloadEnabled || config?.intents;
  const mode = useInteractionModeState();
  const menuIcon = getMenuTriggerIcon(mode, isZoomEnabled);
  const menuLabel = !isPanEnabled && !isZoomEnabled ? intl.formatMessage(TIMESERIES_MESSAGES.zoomModeExplore) : void 0;
  const rawData = useTimeseriesRawData();
  const handleDownloadDataToolbar = () => buildAndDownloadCsv(rawData);
  const menuDisplayMode = getMenuTriggerDisplayMode({
    isPanEnabled,
    isZoomEnabled,
    isDownloadEnabled
  });
  const onClose = () => selectionAreaRef ? selectionAreaRef.current?.focus() : void 0;
  const showDefaultToolbarDownloadButton = isDownloadEnabled && !config?.intents;
  const showMenuToolbarButton = isDownloadEnabled && !!config?.intents?.length;
  const behaviorTrackingProps = _useBehavioralTrackingProps(config ?? {})[0];
  return !isUndefined(config) && anyInteractionEnabled ? /* @__PURE__ */ jsxs(
    ChartToolbarRenderer,
    {
      visible: false,
      handlerRef,
      visibleWhenMouseInBounds: true,
      collapsed: config.collapsed,
      draggable: true,
      placement: config.placement || defaultToolbarPlacement,
      initialA11yState: getA11yToolbarConfig({
        zoomX,
        pan,
        zoom,
        download: config.downloadData,
        intent: !!config.intents
      }),
      zIndex: ChartOverlaysElevation.toolbar,
      onMouseEnter,
      onMouseLeave,
      menuIcon,
      menuTriggerLabel: menuLabel,
      menuDisplayMode,
      onClose,
      isDownloadEnabled: showMenuToolbarButton,
      intents: config?.intents,
      handleDownloadDataToolbar,
      ...behaviorTrackingProps,
      children: [
        /* @__PURE__ */ jsxs(Fragment, { children: [
          /* @__PURE__ */ jsx(ExploreButton, {}),
          isZoomEnabled && /* @__PURE__ */ jsx(ZoomModeButton, {}),
          isPanEnabled && /* @__PURE__ */ jsx(PanButton, {})
        ] }),
        isZoomEnabled && /* @__PURE__ */ jsx(ZoomInOutButtons, {}),
        isZoomEnabled && /* @__PURE__ */ jsx(ResetButton, {}),
        showDefaultToolbarDownloadButton && /* @__PURE__ */ jsx(DownloadDataButton, {})
      ]
    }
  ) : null;
};
TimeseriesChartToolbar["displayName"] = "ToolbarChartToolbar";
export {
  TimeseriesChartToolbar
};
