// @dynatrace/strato-components/esm/forms/spin-buttons/number-spin-button/NumberSpinButton.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import {
  forwardRef,
  useCallback,
  useEffect,
  useImperativeHandle,
  useRef,
  useState
} from "react";
import { isElement } from "react-is";
import { useBeforeInputHandler } from "./hooks/useBeforeInputHandler.js";
import { useValidationAndFormatting } from "./hooks/useValidationAndFormatting.js";
import { useValueNavigation } from "./hooks/useValueNavigation.js";
import { useAriaLabelingProps } from "../../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../../core/hooks/useBehavioralTrackingProps.js";
import { mergeProps } from "../../../core/utils/merge-props.js";
import { useSelection } from "../hooks/useSelection.js";
import { spinButtonCSS } from "../SpinButton.sty.js";
import { NUMBER_SPIN_BUTTON_DEFAULTS } from "./constants.js";
import { useSpinButtonInteraction } from "../hooks/useSpinButtonInteraction.js";
const NumberSpinButton = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const {
    disabled = false,
    readOnly = false,
    value: valueProp,
    placeholder,
    className: consumerClassName,
    style: consumerStyle,
    displayFormat,
    required,
    "data-testid": dataTestId,
    onChange,
    defaultValue = NUMBER_SPIN_BUTTON_DEFAULTS.VALUE,
    minValue,
    maxValue,
    hasError,
    onKeyDownHandler: propKeyDownHandler,
    isCyclic = false,
    "data-pattern": dataPattern,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, remainingPropsWithoutTracking] = useBehavioralTrackingProps(remainingProps);
  const [text, setText] = useState("");
  const [value, setValue] = useState(defaultValue);
  const [error, setError] = useState(false);
  const ariaProps = useAriaLabelingProps(props);
  const rootRef = useRef(null);
  const { getSelection, setSelection, selectAll, deselectAll } = useSelection(rootRef);
  const { validate, format, parse } = useValidationAndFormatting(
    displayFormat,
    required,
    minValue ?? NUMBER_SPIN_BUTTON_DEFAULTS.MIN_VALUE,
    maxValue ?? NUMBER_SPIN_BUTTON_DEFAULTS.MAX_VALUE
  );
  const applyText = useCallback(
    (input, fireOnChange = false) => {
      const newValue = parse(input);
      setText(input);
      setValue(newValue);
      if (rootRef?.current) {
        rootRef.current.textContent = format(input);
      }
      fireOnChange && onChange?.(newValue);
    },
    [format, onChange, parse, setValue]
  );
  const clearValue = useCallback(
    (element, fireOnChange = false) => {
      setValue(null);
      setText("");
      if (element) {
        element.textContent = "";
      }
      fireOnChange && onChange?.(null);
    },
    [onChange, setValue]
  );
  const beforeInputHandler = useBeforeInputHandler(
    rootRef,
    props,
    applyText,
    text
  );
  const { onKeyDownHandler } = useValueNavigation(
    rootRef,
    {
      value: valueProp ?? value,
      disabled,
      readOnly,
      onChange,
      displayFormat,
      required,
      minValue,
      maxValue,
      isCyclic,
      ...remainingPropsWithoutTracking
    },
    applyText
  );
  useImperativeHandle(
    forwardedRef,
    () => ({
      type: "number",
      element: rootRef,
      setSelection,
      getSelection,
      selectAll,
      deselectAll,
      applyText: (input, fireOnChange) => applyText(input, fireOnChange),
      clearValue: (fireOnChange) => clearValue(rootRef.current, fireOnChange)
    }),
    [setSelection, getSelection, selectAll, deselectAll, clearValue, applyText]
  );
  useEffect(() => {
    const isValid = validate(value);
    setError(!isValid);
  }, [setError, validate, value]);
  useEffect(() => {
    applyText(
      defaultValue === null || defaultValue === void 0 || Number.isNaN(defaultValue) ? "" : defaultValue.toString()
    );
  }, [format]);
  useEffect(() => {
    const element = rootRef.current;
    element?.addEventListener("beforeinput", beforeInputHandler);
    return () => element?.removeEventListener("beforeinput", beforeInputHandler);
  }, [beforeInputHandler]);
  useEffect(() => {
    if (valueProp === void 0) {
      return;
    }
    const sel = getSelection();
    if (valueProp !== value) {
      applyText(
        valueProp !== null && Number.isInteger(valueProp) ? valueProp.toString() : ""
      );
    }
    document.activeElement === rootRef.current && sel && setSelection(sel);
  }, [valueProp, applyText, format, getSelection, setSelection]);
  const { onClick, onFocus, onBlur } = useSpinButtonInteraction({
    disabled,
    selectAll,
    deselectAll,
    rootElement: rootRef.current
  });
  return /* @__PURE__ */ jsx(
    "span",
    {
      "data-testid": dataTestId,
      ref: rootRef,
      tabIndex: disabled ? void 0 : 0,
      contentEditable: (
        // contentEditable when focused and non-editable when blurred, allowing text selection.
        !disabled && !readOnly ? true : void 0
      ),
      spellCheck: false,
      className: clsx(
        consumerClassName,
        spinButtonCSS({ disabled, error: hasError ?? error })
      ),
      style: {
        "--data-placeholder": `"${placeholder ?? ""}"`,
        ...consumerStyle
      },
      "aria-disabled": disabled,
      ...mergeProps(ariaProps, behavioralTrackingProps),
      "aria-label": ariaProps["aria-label"] ?? placeholder,
      role: "spinbutton",
      onClick,
      onFocus,
      onBlur,
      onKeyDown: propKeyDownHandler ?? onKeyDownHandler,
      "data-pattern": dataPattern
    }
  );
});
function isNumberSpinButton(node) {
  return node !== void 0 && node !== null && isElement(node) && node.type === NumberSpinButton;
}
NumberSpinButton.displayName = "_NumberSpinButton";
export {
  NumberSpinButton,
  isNumberSpinButton
};
