// @dynatrace/strato-components/content/ai-response/AiResponse.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var AiResponse_exports = {};
__export(AiResponse_exports, {
  AiResponse: () => AiResponse
});
module.exports = __toCommonJS(AiResponse_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_framer_motion = require("framer-motion");
var import_react = require("react");
var import_streamdown = require("streamdown");
var styles = __toESM(require("./AiResponse.sty.js"));
var import_useTextAnimation = require("./hooks/useTextAnimation.js");
var import_component_mappings = require("./utils/component-mappings.js");
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useSafeForwardProps = require("../../core/hooks/useSafeForwardProps.js");
const COMPONENT_NAME = "AiResponse";
const animationOptions = {
  animation: "aiResponseAnimation",
  duration: import_useTextAnimation.ANIMATION_DURATION_MS,
  easing: "cubic-bezier(0, 0, 0.2, 1)",
  sep: "word"
};
const InternalAiResponse = (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const {
      children,
      onAnimationStateChange,
      responseState = "streaming",
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
    const ariaLabelingProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(restProps);
    const forwardProps = (0, import_useSafeForwardProps.useSafeForwardProps)(restProps, ariaLabelingProps);
    const shouldReduceMotion = (0, import_framer_motion.useReducedMotion)() ?? false;
    const { displayedText, isTextAnimating } = (0, import_useTextAnimation.useTextAnimation)(
      children,
      shouldReduceMotion,
      responseState,
      onAnimationStateChange
    );
    if (displayedText.length === 0) {
      return null;
    }
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "div",
      {
        ref: forwardedRef,
        "aria-live": "polite",
        "aria-atomic": "true",
        style: consumerStyle,
        className: (0, import_clsx.default)(styles.wrapper, consumerClassName),
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        ...ariaLabelingProps,
        ...behavioralTrackingProps,
        ...forwardProps,
        "data-dt-component": COMPONENT_NAME,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_streamdown.Streamdown,
          {
            animated: shouldReduceMotion ? false : animationOptions,
            isAnimating: responseState === "streaming" || isTextAnimating,
            components: import_component_mappings.streamdownComponents,
            className: styles.streamdown,
            children: displayedText
          }
        )
      }
    );
  }
);
InternalAiResponse.displayName = COMPONENT_NAME;
const AiResponse = (0, import_react.memo)(InternalAiResponse);
