// @dynatrace/strato-components/charts/xy-chart-adoption/timeseries/TimeseriesToXY.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TimeseriesToXY_exports = {};
__export(TimeseriesToXY_exports, {
  TimeseriesToXY: () => TimeseriesToXY
});
module.exports = __toCommonJS(TimeseriesToXY_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_iterate_slots = require("./iterate-slots.js");
var import_transformer_data_to_xy = require("./transformer-data-to-xy.js");
var import_timeseries = require("./types/timeseries.js");
var import_render_series = require("./utils/render-series.js");
var import_XYChartRoot = require("../../xy-chart/XYChartRoot.js");
const TimeseriesToXY = (0, import_react.forwardRef)((props, _ref) => {
  const { data = [], variant = import_timeseries.DEFAULT_VARIANT, children } = props;
  const { line, bar, area, band, shapePriority } = (0, import_iterate_slots.iterateSeriesSlots)(children);
  const buckets = {
    line: variant === "line" ? [...data, ...line] : line,
    area: variant === "area" ? [...data, ...area] : area,
    bar: variant === "bar" ? [...data, ...bar] : bar,
    band
  };
  const order = [...shapePriority];
  if (data.length > 0 && !order.includes(variant)) {
    order.push(variant);
  }
  const result = (0, import_transformer_data_to_xy.transformTimeseriesDataToXYChart)(buckets, order);
  const seriesForShape = (shape) => {
    switch (shape) {
      case "line":
        return result.line.map(import_render_series.renderLineSeries);
      case "area":
        return result.area.map(import_render_series.renderAreaSeries);
      case "bar":
        return result.bar.map(import_render_series.renderBarSeries);
      case "band":
        return result.band.map(import_render_series.renderBandSeries);
    }
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_XYChartRoot.XYChart, { data: [], children: [
    order.flatMap(seriesForShape),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_XYChartRoot.XYChart.XAxis,
      {
        id: import_render_series.xAxisId,
        type: "time",
        position: "bottom",
        min: "data-min",
        max: "data-max"
      }
    ),
    [...result.units].map((unit, index) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_XYChartRoot.XYChart.YAxis,
      {
        id: (0, import_render_series.defineAxisId)(unit),
        type: "numerical",
        unit,
        position: index % 2 === 0 ? "left" : "right"
      },
      (0, import_render_series.defineAxisId)(unit)
    ))
  ] });
});
