// @dynatrace/strato-components/charts/xy-chart-adoption/timeseries/transformer-data-to-xy.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var transformer_data_to_xy_exports = {};
__export(transformer_data_to_xy_exports, {
  transformTimeseriesDataToXYChart: () => transformTimeseriesDataToXYChart
});
module.exports = __toCommonJS(transformer_data_to_xy_exports);
var import_single_line_name = require("../../core/utils/single-line-name.js");
const mapTimeseries = (series, seriesId) => {
  const { datapoints, name, unit, ...extraSeriesProps } = series;
  const seriesName = (0, import_single_line_name.singleLineName)(name);
  const data = datapoints.map((datapoint) => ({
    ...extraSeriesProps,
    start: datapoint.start,
    end: datapoint.end,
    center: datapoint.center,
    value: datapoint.value,
    seriesName,
    seriesId
  }));
  return { data, unit, seriesId };
};
const mapBand = (band, seriesId) => {
  const { datapoints, name, unit, ...extraSeriesProps } = band;
  const seriesName = (0, import_single_line_name.singleLineName)(name);
  const data = datapoints.map((datapoint) => ({
    ...extraSeriesProps,
    start: datapoint.start,
    end: datapoint.end,
    y0: datapoint.y0,
    y1: datapoint.y1,
    seriesName,
    seriesId
  }));
  return { data, unit, seriesId };
};
function transformTimeseriesDataToXYChart(input, shapeRenderOrder) {
  const units = /* @__PURE__ */ new Set();
  const result = {
    line: [],
    area: [],
    bar: [],
    band: [],
    units
  };
  let globalIndex = 0;
  shapeRenderOrder.forEach((shape) => {
    if (shape === "band") {
      input.band.forEach((band) => {
        const mapped = mapBand(band, `${shape}-${globalIndex}`);
        result.band.push(mapped);
        units.add(mapped.unit);
        globalIndex++;
      });
    } else {
      input[shape].forEach((series) => {
        const mapped = mapTimeseries(series, `${shape}-${globalIndex}`);
        result[shape].push(mapped);
        units.add(mapped.unit);
        globalIndex++;
      });
    }
  });
  return result;
}
