// @dynatrace/strato-components/esm/tables/DataTable/features/Layout/layout-utils.js
import { isFractionWidth, isWidthExtendedConfig } from "./width-utils.js";
import { devLog } from "../../../../core/utils/logging.js";
import {
  DATA_TABLE_CELL_VERTICAL_PADDING_MAP,
  DATA_TABLE_DEFAULT_ICON_PADDING,
  DATA_TABLE_DEFAULT_ICON_SIZE,
  DATA_TABLE_DEFAULT_MIN_COLUMN_WIDTH,
  DATA_TABLE_STRATO_RESIZING_SIZE_CSS_VAR,
  DATA_TABLE_THRESHOLD_PILL_SIZE
} from "../../constants.js";
import { isRowSelectionColumnDef } from "../RowSelection/RowSelection.js";
function getPrefixColumnWidth(column, rowDensity, rowThresholds) {
  const isFirstColumn = column.getIsFirstColumn();
  const iconPadding = isRowSelectionColumnDef(column.columnDef) ? 0 : 2 * DATA_TABLE_DEFAULT_ICON_PADDING;
  let rowDensityPadding = DATA_TABLE_CELL_VERTICAL_PADDING_MAP[rowDensity];
  let thresholdPillSize = 0;
  if (isFirstColumn) {
    rowDensityPadding *= 2;
    const hasThresholdPill = rowThresholds?.some(
      (threshold) => threshold.type === "pill"
    );
    thresholdPillSize = hasThresholdPill ? DATA_TABLE_THRESHOLD_PILL_SIZE : 0;
  }
  return DATA_TABLE_DEFAULT_ICON_SIZE + iconPadding + thresholdPillSize + rowDensityPadding;
}
function getTargetTemplate(pinnedState, templates) {
  if (!pinnedState) {
    return templates.center;
  }
  return templates[pinnedState];
}
function clampWidthToPx(width, minWidth, maxWidth) {
  return `${Math.min(Math.max(width, minWidth), maxWidth)}px`;
}
function calculateParentDependentWidth(column, parentWidth, getMeasuredWidth) {
  let totalDefinedChildWidths = 0;
  let undefinedWidthChildCount = 1;
  column.parent?.getVisibleLeafColumns().forEach((childCol) => {
    const childWidth = childCol.columnDef.width;
    if (childWidth === void 0 && childCol.id !== column.id) {
      undefinedWidthChildCount++;
    } else if (typeof childWidth === "number") {
      totalDefinedChildWidths += childWidth;
    } else if (childWidth === "auto" || childWidth === "content") {
      totalDefinedChildWidths += getMeasuredWidth(childCol.id) ?? 0;
    } else if (isFractionWidth(childWidth)) {
      devLog(
        "Fraction width is not allowed in combination with a fixed width parent column."
      );
    }
  });
  const remainingWidth = (parentWidth - totalDefinedChildWidths) / undefinedWidthChildCount;
  return `${remainingWidth}px`;
}
function getFractionWidthTemplateValue(parent, width, minWidth) {
  if (!parent || !isFractionWidth(parent.columnDef.width)) {
    return `minmax(${minWidth}px, ${width})`;
  }
  const parentFraction = Number.parseFloat(parent.columnDef.width);
  const columnFraction = Number.parseFloat(width);
  const visibleSiblingColumns = parent.getVisibleLeafColumns();
  const siblingColumnsSum = visibleSiblingColumns.reduce((sum, col) => {
    return isFractionWidth(col.columnDef.width) ? sum + Number.parseFloat(col.columnDef.width) : sum;
  }, 0);
  const calculatedFraction = parentFraction * (columnFraction / siblingColumnsSum);
  return `${calculatedFraction}fr`;
}
function getContentOrAutoWidthTemplateValue(measuredWidth, isContent) {
  if (measuredWidth !== void 0) {
    return `${measuredWidth}px`;
  }
  return isContent ? "max-content" : "minmax(min-content, auto)";
}
function getStringWidthTemplateValue(column, width, minWidth, getMeasuredWidth) {
  if (isFractionWidth(width)) {
    return getFractionWidthTemplateValue(column.parent, width, minWidth);
  }
  if (isWidthExtendedConfig(width)) {
    return getContentOrAutoWidthTemplateValue(
      getMeasuredWidth(column.id),
      width.type === "content"
    );
  }
  return getContentOrAutoWidthTemplateValue(
    getMeasuredWidth(column.id),
    width === "content"
  );
}
function getUndefinedWidthTemplateValue(column, minWidth, getMeasuredWidth) {
  const parentWidth = column.parent?.columnDef.width;
  if (column.parent && typeof parentWidth === "number") {
    return calculateParentDependentWidth(column, parentWidth, getMeasuredWidth);
  }
  const measuredWidth = getMeasuredWidth(column.id, true);
  return measuredWidth === void 0 ? `minmax(${minWidth}px, 1fr)` : `${measuredWidth}px`;
}
function processBuiltinColumn(column, rowDensity, rowThresholds, columnGridTemplates) {
  if (column.columnDef.meta?.builtinPlacement === "prefix") {
    columnGridTemplates.left.push(
      `${getPrefixColumnWidth(column, rowDensity, rowThresholds)}px`
    );
    return true;
  }
  if (column.columnDef.meta?.spacer === true) {
    columnGridTemplates.center.push("auto");
    return true;
  }
  if (column.columnDef.meta?.builtinPlacement === "suffix") {
    columnGridTemplates.right.push("min-content");
    return true;
  }
  return false;
}
function getColumnTemplateEntry(column, columnSizing, columnSizingInfo, getMeasuredWidth) {
  const {
    minWidth = DATA_TABLE_DEFAULT_MIN_COLUMN_WIDTH,
    width,
    maxWidth = Number.MAX_SAFE_INTEGER
  } = column.columnDef;
  if (columnSizingInfo.isResizingColumn === column.id) {
    return `var(${DATA_TABLE_STRATO_RESIZING_SIZE_CSS_VAR})`;
  }
  if (columnSizing[column.id] !== void 0) {
    return clampWidthToPx(columnSizing[column.id], minWidth, maxWidth);
  }
  if (width === void 0) {
    return getUndefinedWidthTemplateValue(column, minWidth, getMeasuredWidth);
  }
  if (typeof width === "number") {
    return clampWidthToPx(width, minWidth, maxWidth);
  }
  if (typeof width === "string" || isWidthExtendedConfig(width)) {
    return getStringWidthTemplateValue(
      column,
      width,
      minWidth,
      getMeasuredWidth
    );
  }
  return void 0;
}
export {
  getColumnTemplateEntry,
  getPrefixColumnWidth,
  getTargetTemplate,
  processBuiltinColumn
};
