// @dynatrace/strato-components/esm/filters/filter-field/utils/conversion/lezer-tree-conversion/key/key.js
import { convertJsonPath } from "./json-path.js";
import { isDev } from "../../../../../../core/utils/is-node-environment.js";
import { preSpotErrorMark } from "../../../../decorations/decoration-marks.js";
import { getTextFromRange } from "../../../get-text.js";
import { convertNode } from "../generic-node.js";
import { updateNodeValues, convertToErrorNode } from "../node-mutations.js";
import { convertIdentifier } from "../value/identifier.js";
function convertFilterKey(node, doc, parserConfig) {
  const { from, to } = node;
  const cursor = node.node.cursor();
  let convertedNode = {
    type: "Key",
    range: {
      from,
      to
    },
    textValue: "",
    value: ""
  };
  const errors = [];
  let prevNode;
  cursor.iterate((childNode) => {
    const { from: childFrom, to: childTo } = childNode;
    const nodeName = childNode.name;
    switch (nodeName) {
      case "FilterKey": {
        break;
      }
      case "Identifier":
      case "String":
      case "IncompleteString":
      case "Variable": {
        if (prevNode === "JsonPathSquareOpeningBracket") {
          break;
        }
        const {
          convertedNode: { textValue, value },
          isValid: isValidIdentifier,
          positionOperators: { endsWithOperator, startsWithOperator }
        } = convertIdentifier(childNode, doc, nodeName);
        updateNodeValues(convertedNode, textValue, value);
        if (!isValidIdentifier || nodeName === "IncompleteString" || endsWithOperator || startsWithOperator) {
          convertToErrorNode(convertedNode);
          errors.push({ range: { from, to }, message: "invalidKey" });
        }
        if (endsWithOperator) {
          errors.push({
            range: endsWithOperator,
            message: "unsupportedPositionOperator"
          });
        }
        if (startsWithOperator) {
          errors.push({
            range: startsWithOperator,
            message: "unsupportedPositionOperator"
          });
        }
        break;
      }
      case "InvalidIdentifier": {
        convertedNode = convertNode(childNode, doc).convertedNode;
        errors.push({
          range: { from: childFrom, to: childTo },
          message: "invalidKey"
        });
        break;
      }
      case "PositionOperator":
      case "InvalidPositionOperator": {
        const text = getTextFromRange(childNode, doc);
        updateNodeValues(convertedNode, text);
        convertToErrorNode(convertedNode);
        errors.push({
          range: { from: childFrom, to: childTo },
          message: "unsupportedPositionOperator"
        });
        break;
      }
      case "JsonPath": {
        const { convertedNode: jsonPathNode, errors: jsonPathErrors } = convertJsonPath(childNode, doc, parserConfig);
        convertedNode = jsonPathNode;
        errors.push(...jsonPathErrors);
        cursor.lastChild();
        break;
      }
      case "\u26A0":
        if (prevNode !== "PositionOperator" && prevNode !== "InvalidPositionOperator") {
          if (prevNode !== "Identifier") {
            errors.push({
              range: { from: childFrom, to: childTo + 1 },
              message: "missingKeyError",
              decoration: preSpotErrorMark
            });
          } else {
            const textValue = getTextFromRange(childNode, doc);
            if (textValue === ")" || textValue === "]") {
              updateNodeValues(convertedNode, textValue);
              convertToErrorNode(convertedNode);
              errors.push({
                range: { from, to: childTo },
                message: "invalidKey"
              });
            }
          }
        }
        break;
      case "JsonPathSquareOpeningBracket":
        break;
      default:
        if (isDev()) {
          console.log(
            `${doc.toString()} | Syntax node '${childNode.name}' is not yet handled in the key conversion. Please make sure to handle all nodes accordingly.`
          );
        }
    }
    prevNode = nodeName;
  });
  if (convertedNode.type === "Error") {
    convertedNode.value = convertedNode.textValue;
  }
  return {
    // For an empty error node, we return undefined.
    convertedNode: from === to ? void 0 : convertedNode,
    errors
  };
}
export {
  convertFilterKey
};
