// @dynatrace/strato-components/esm/tables/DataTable/features/VisualVariants/VisualVariants.js
import { isDev } from "../../../../core/utils/is-node-environment.js";
import { useDataTableConfigElement } from "../../../hooks/useDataTableConfig.js";
import { DATA_TABLE_VISUAL_VARIANT_CONFIG_DEFAULT } from "../../constants.js";
import { memo } from "../../tanstack-table-core-override/src/index.js";
function isVerticalAlignmentExtendedConfig(verticalAlignment) {
  if (verticalAlignment !== void 0 && typeof verticalAlignment !== "string") {
    return true;
  }
  return false;
}
function useVisualVariant(props, options) {
  const configVariant = useDataTableConfigElement("variant");
  options.state = {
    ...options.state,
    visualVariantConfig: {
      rowSeparation: props.variant?.rowSeparation ?? configVariant?.rowSeparation ?? DATA_TABLE_VISUAL_VARIANT_CONFIG_DEFAULT.rowSeparation,
      verticalDividers: props.variant?.verticalDividers ?? configVariant?.verticalDividers ?? DATA_TABLE_VISUAL_VARIANT_CONFIG_DEFAULT.verticalDividers,
      contained: props.variant?.contained ?? configVariant?.contained ?? DATA_TABLE_VISUAL_VARIANT_CONFIG_DEFAULT.contained,
      rowDensity: props.variant?.rowDensity ?? configVariant?.rowDensity ?? DATA_TABLE_VISUAL_VARIANT_CONFIG_DEFAULT.rowDensity,
      fontStyle: props.variant?.fontStyle ?? configVariant?.fontStyle ?? DATA_TABLE_VISUAL_VARIANT_CONFIG_DEFAULT.fontStyle,
      headers: props.variant?.headers ?? configVariant?.headers ?? DATA_TABLE_VISUAL_VARIANT_CONFIG_DEFAULT.headers,
      verticalAlignment: props.variant?.verticalAlignment ?? configVariant?.verticalAlignment ?? DATA_TABLE_VISUAL_VARIANT_CONFIG_DEFAULT.verticalAlignment
    }
  };
}
const VisualVariantConfig = {
  getInitialState(state) {
    return {
      visualVariantConfig: {
        rowSeparation: DATA_TABLE_VISUAL_VARIANT_CONFIG_DEFAULT.rowSeparation,
        verticalDividers: DATA_TABLE_VISUAL_VARIANT_CONFIG_DEFAULT.verticalDividers,
        contained: DATA_TABLE_VISUAL_VARIANT_CONFIG_DEFAULT.contained,
        rowDensity: DATA_TABLE_VISUAL_VARIANT_CONFIG_DEFAULT.rowDensity,
        fontStyle: DATA_TABLE_VISUAL_VARIANT_CONFIG_DEFAULT.fontStyle,
        headers: DATA_TABLE_VISUAL_VARIANT_CONFIG_DEFAULT.headers,
        verticalAlignment: DATA_TABLE_VISUAL_VARIANT_CONFIG_DEFAULT.verticalAlignment
      },
      ...state
    };
  },
  createTable(table) {
    table.getVisualVariantOptions = () => {
      return table.getState().visualVariantConfig;
    };
    table.getBodyVerticalAlignment = memo(
      () => [table.getVisualVariantOptions().verticalAlignment],
      (verticalAlignment) => {
        if (isVerticalAlignmentExtendedConfig(verticalAlignment)) {
          return verticalAlignment["body"];
        }
        return verticalAlignment;
      },
      { key: isDev() && "getBodyVerticalAlignment" }
    );
    table.getHeaderVerticalAlignment = memo(
      () => [table.getVisualVariantOptions().verticalAlignment],
      (verticalAlignment) => {
        if (isVerticalAlignmentExtendedConfig(verticalAlignment)) {
          return verticalAlignment["header"];
        }
        return verticalAlignment;
      },
      { key: isDev() && "getHeaderVerticalAlignment" }
    );
  }
};
export {
  VisualVariantConfig,
  useVisualVariant
};
