// @dynatrace/strato-components/charts/tree-map/utils/build-overlay-tooltip-data.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var build_overlay_tooltip_data_exports = {};
__export(build_overlay_tooltip_data_exports, {
  buildOverlayTooltipData: () => buildOverlayTooltipData
});
module.exports = __toCommonJS(build_overlay_tooltip_data_exports);
var import_get_tooltip_item = require("./get-tooltip-item.js");
const buildOverlayTooltipData = (node, formatter, unit) => {
  const item = (0, import_get_tooltip_item.getTooltipItem)(formatter, node, unit);
  const parentName = node.parent?.data.name;
  const nodeColor = node.parent?.data.color?.Default ?? node.data?.color?.Default ?? "";
  const path = [...node.ancestors()].reverse().slice(1).map((n) => n.data.name);
  const children = node.children?.map((c) => c.data.name);
  return {
    geometry: "treemap",
    header1: parentName ? [parentName] : [],
    metadata: { id: item.id, value: node.data.value, path, children },
    series: [
      {
        name: item.name,
        color: nodeColor,
        value: item.values[0]
      }
    ]
  };
};
