// @dynatrace/strato-components/charts/tree-map/utils/iterate-slots.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var iterate_slots_exports = {};
__export(iterate_slots_exports, {
  iterateSlots: () => iterateSlots
});
module.exports = __toCommonJS(iterate_slots_exports);
var import_react = require("react");
var import_react_is = require("react-is");
var import_Intent = require("../../../core/slots/Intent/Intent.js");
var import_is_slot = require("../../../core/utils/is-slot.js");
var import_ColorRule = require("../../core/slots/ColorRule.js");
var import_ChartToolbar = require("../../core/slots/toolbar/ChartToolbar.js");
var import_CompactToolbar = require("../../core/slots/toolbar/CompactToolbar.js");
var import_DownloadData = require("../../core/slots/toolbar/DownloadData.js");
var import_extract_behavior_tracking_props = require("../../core/utils/extract-behavior-tracking-props.js");
var import_DownloadCSV = require("../slots/DownloadCSV.js");
var import_Toolbar = require("../slots/Toolbar.js");
var import_Tooltip = require("../slots/Tooltip.js");
var import_TreeMapChartLegend = require("../slots/TreeMapChartLegend.js");
const iterateChartToolbarSlots = (children) => {
  let slots = {};
  import_react.Children.forEach(children, (childInteractions) => {
    if ((0, import_react_is.isElement)(childInteractions)) {
      switch (childInteractions.type) {
        // eslint-disable-next-line @typescript-eslint/no-deprecated
        case import_DownloadData.DownloadData:
          slots = { ...slots, downloadData: { enabled: true } };
          break;
      }
    }
  });
  return slots;
};
function iterateSlots(children) {
  let slots = {
    legend: {},
    tooltip: void 0,
    toolbar: void 0,
    colorRules: []
  };
  let intents;
  import_react.Children.map(children, (child) => {
    if ((0, import_is_slot.isSlot)(child, import_Tooltip.Tooltip)) {
      slots = {
        ...slots,
        tooltip: {
          hidden: child.props?.hidden ?? false,
          template: child.props?.children
        }
      };
    } else if ((0, import_is_slot.isSlot)(child, import_TreeMapChartLegend.Legend)) {
      const legend = child.props;
      slots = { ...slots, legend };
    } else if ((0, import_is_slot.isSlot)(child, import_ChartToolbar.ChartToolbar)) {
      const { children: children2, ...toolbarProps } = child.props;
      const toolbarSlots = iterateChartToolbarSlots(children2);
      const toolbar = { ...toolbarProps, ...toolbarSlots };
      slots = { ...slots, toolbar };
    } else if ((0, import_is_slot.isSlot)(child, import_ColorRule.ColorRule)) {
      const colorRule = child.props;
      if (slots.colorRules) {
        slots.colorRules.push(colorRule);
      }
    } else if ((0, import_is_slot.isSlot)(child, import_Intent.Intent)) {
      intents = intents ? [...intents, child.props] : [child.props];
    } else if ((0, import_is_slot.isSlot)(child, import_Toolbar.Toolbar)) {
      slots = {
        ...slots,
        toolbar: {
          ...slots.toolbar,
          ...(0, import_CompactToolbar.sanitizeProps)(child.props),
          ...(0, import_extract_behavior_tracking_props.extractBehaviorTrackingProps)(child.props)
        }
      };
    } else if ((0, import_is_slot.isSlot)(child, import_DownloadCSV.TreeMapDownloadCSVSlot)) {
      slots = {
        ...slots,
        toolbar: {
          ...slots.toolbar,
          downloadData: { enabled: true }
        }
      };
    }
  });
  if (intents?.length) {
    slots = {
      ...slots,
      toolbar: {
        ...slots.toolbar,
        intents
      }
    };
  }
  return slots;
}
