// @dynatrace/strato-components/charts/tree-map/utils/node-aggregations.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var node_aggregations_exports = {};
__export(node_aggregations_exports, {
  getAggregatedCluster: () => getAggregatedCluster,
  getAggregatedLeaf: () => getAggregatedLeaf
});
module.exports = __toCommonJS(node_aggregations_exports);
var import_get_nodes_value_sum = require("./get-nodes-value-sum.js");
var import_constants = require("../constants.js");
const aggregatedColorToken = {
  Default: import_constants.AGGREGATED_NODE_COLOR
};
const getAggregatedLeaf = (nodes, clusterName, formatter, clusterColorToken, nameAccessor) => {
  const aggeregatedValue = (0, import_get_nodes_value_sum.getNodesValueSum)(nodes);
  return {
    id: `other-${clusterName}`,
    value: aggeregatedValue,
    formattedValue: formatter(aggeregatedValue),
    name: import_constants.DEFAULT_AGGREGATED_LABEL,
    color: nameAccessor ? aggregatedColorToken : clusterColorToken,
    ...nameAccessor && {
      aggregatedNodes: nodes.map((node) => node[nameAccessor])
    }
  };
};
const getAggregatedCluster = (nodes, formatter) => ({
  id: import_constants.CLUSTER_OTHER_ID,
  value: (0, import_get_nodes_value_sum.getNodesValueSum)(nodes),
  name: import_constants.DEFAULT_AGGREGATED_LABEL,
  color: aggregatedColorToken,
  // for highlighting logic, the generated id must be 'other-cluster', same as CLUSTER_OTHER_ID
  children: [
    getAggregatedLeaf(nodes, "cluster", formatter, aggregatedColorToken)
  ],
  aggregatedNodes: nodes.map((node) => node.name)
});
