// @dynatrace/strato-components/esm/forms/switch/Switch.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import {
  forwardRef,
  useImperativeHandle,
  useRef
} from "react";
import { CheckmarkIcon, XmarkIcon } from "@dynatrace/strato-icons";
import {
  switchIconCSS,
  switchLabelCSS,
  switchTextCSS,
  switchThumbCSS,
  switchWrapperCSS
} from "./Switch.sty.js";
import { VisuallyHidden } from "../../core/components/visually-hidden/VisuallyHidden.js";
import { useAriaLabelingProps as _useAriaLabelingProps } from "../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useControllableState } from "../../core/hooks/useControllableState.js";
import { useFormControlState } from "../../core/hooks/useFormControlState.js";
import { useFocusRing } from "../../core/styles/useFocusRing.js";
import { isDev } from "../../core/utils/is-node-environment.js";
import { mergeProps as _mergeProps } from "../../core/utils/merge-props.js";
import { Flex } from "../../layouts/flex/Flex.js";
import { fieldTokens as fieldColorsCSS } from "../../styles/field.sty.js";
import { InternalBaseItem as BaseItem } from "../forms-core/components/form-field-messages/FormFieldMessageBaseItem.js";
import { useFormFieldMessagesContext } from "../forms-core/hooks/useFormFieldMessagesContext.js";
import { useFormFieldValidation } from "../forms-core/hooks/useFormFieldValidation.js";
import { useSyncedFormControlProps } from "../forms-core/hooks/useSyncedFormControlProps.js";
import { FormFieldMessagesAction } from "../forms-core/state/formfield-messages-state-reducer.js";
import { useInputHint } from "../hooks/useInputHint.js";
const COMPONENT_NAME = "Switch";
const Switch = forwardRef(
  (props, forwardedRef) => {
    const {
      value,
      defaultValue,
      children,
      controlState,
      name,
      formValue,
      onChange,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      id: propId,
      required: propRequired,
      disabled: propDisabled,
      "aria-disabled": ariaDisabledProp,
      readOnly: propReadOnly,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps] = useBehavioralTrackingProps(remainingProps);
    const { state: messagesContextState, dispatch: messagesContextDispatch } = useFormFieldMessagesContext();
    const { disabled, required, id, ariaDisabled, readOnly } = useSyncedFormControlProps({
      propDisabled,
      propRequired,
      propId,
      propAriaDisabled: ariaDisabledProp,
      propReadOnly
    });
    const disabledOrAriaDisabled = disabled || ariaDisabled || false;
    const ariaLabelingProps = _useAriaLabelingProps(props);
    const rootRef = useRef(null);
    const inputRef = useRef(null);
    useImperativeHandle(forwardedRef, () => {
      return {
        element: rootRef.current,
        inputRef: inputRef.current,
        focus() {
          inputRef.current?.focus();
        },
        validate() {
          inputRef.current?.checkValidity();
          messagesContextDispatch?.({
            type: FormFieldMessagesAction.SETISTOUCHED,
            payload: { isTouched: true }
          });
        }
      };
    });
    const {
      onChange: validationOnChange,
      onBlur: validationOnBlur,
      onInvalid: validationOnInvalid,
      ariaProps
    } = useFormFieldValidation(inputRef, { required });
    const [controlStateHasError, hint] = useFormControlState(controlState);
    const showError = controlStateHasError || messagesContextState.showError;
    const { inputProps, hintProps } = useInputHint({
      hint,
      hasError: showError
    });
    if (value !== void 0 && !onChange && isDev()) {
      console.log(
        'Defining "value" makes the switch controlled. Set the onChange() callback and implement the toggle functionality.'
      );
    }
    const [isSelected, setIsSelected] = useControllableState({
      value,
      defaultValue
    });
    const color = isSelected ? "primary" : "neutral";
    let focusRingColor = "neutral";
    if (showError) {
      focusRingColor = "critical";
    } else if (!(readOnly || disabledOrAriaDisabled) && isSelected) {
      focusRingColor = "primary";
    }
    const { focusClassName, focusProps, isFocusVisible } = useFocusRing({
      color: focusRingColor,
      disabled
    });
    function handleChange(e) {
      if (!disabled && !readOnly) {
        onChange?.(!isSelected, e);
        setIsSelected(!isSelected);
        !controlState && validationOnChange();
      }
    }
    return /* @__PURE__ */ jsxs(
      Flex,
      {
        ref: rootRef,
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        flexDirection: "column",
        gap: 4,
        className: consumerClassName,
        style: consumerStyle,
        ...behavioralTrackingProps,
        "data-dt-component": COMPONENT_NAME,
        children: [
          /* @__PURE__ */ jsxs(
            "label",
            {
              htmlFor: id,
              className: clsx(
                switchLabelCSS({ disabled: disabledOrAriaDisabled, readOnly })
              ),
              children: [
                /* @__PURE__ */ jsx(VisuallyHidden, { children: /* @__PURE__ */ jsx(
                  "input",
                  {
                    id,
                    type: "checkbox",
                    role: "switch",
                    checked: isSelected,
                    "aria-disabled": ariaDisabled,
                    disabled,
                    required,
                    readOnly,
                    "aria-checked": isSelected,
                    onChange: handleChange,
                    onBlur: (event) => {
                      focusProps.onBlur(event);
                      !controlState && validationOnBlur(event);
                    },
                    onFocus: focusProps.onFocus,
                    onInvalid: (event) => {
                      !controlState && validationOnInvalid(event);
                    },
                    name,
                    value: formValue,
                    ref: inputRef,
                    ..._mergeProps(
                      controlState ? inputProps : ariaProps,
                      ariaLabelingProps
                    )
                  }
                ) }),
                /* @__PURE__ */ jsx(
                  "div",
                  {
                    className: clsx(
                      focusClassName,
                      switchWrapperCSS({
                        disabled: disabledOrAriaDisabled,
                        color,
                        hasError: showError,
                        isFocusVisible,
                        readOnly
                      }),
                      fieldColorsCSS({
                        disabled: disabledOrAriaDisabled,
                        color: showError ? "critical" : color,
                        variant: isSelected ? "accent" : showError ? "emphasized" : "default"
                      })
                    ),
                    children: /* @__PURE__ */ jsxs(
                      "div",
                      {
                        className: switchIconCSS({
                          selected: isSelected,
                          readOnly,
                          hasError: showError
                        }),
                        children: [
                          /* @__PURE__ */ jsx(CheckmarkIcon, { size: "small" }),
                          /* @__PURE__ */ jsx("div", { className: switchThumbCSS }),
                          /* @__PURE__ */ jsx(XmarkIcon, { size: "small" })
                        ]
                      }
                    )
                  }
                ),
                /* @__PURE__ */ jsx(
                  "span",
                  {
                    className: switchTextCSS({
                      disabled: disabledOrAriaDisabled,
                      required: !!children && required
                    }),
                    children
                  }
                )
              ]
            }
          ),
          hint && /* @__PURE__ */ jsx(
            BaseItem,
            {
              variant: showError ? "error" : "hint",
              disabled: disabledOrAriaDisabled,
              ...hintProps,
              children: hint
            }
          )
        ]
      }
    );
  }
);
Switch.displayName = COMPONENT_NAME;
export {
  Switch
};
