// @dynatrace/strato-components/forms/forms-core/components/form-field/FormFieldMessageContextProvider.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var FormFieldMessageContextProvider_exports = {};
__export(FormFieldMessageContextProvider_exports, {
  FormFieldMessageContextProvider: () => FormFieldMessageContextProvider,
  withFormFieldMessageContextProvider: () => withFormFieldMessageContextProvider,
  withFormFieldMessageContextProviderWithGenerics: () => withFormFieldMessageContextProviderWithGenerics
});
module.exports = __toCommonJS(FormFieldMessageContextProvider_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_useId = require("../../../../core/hooks/useId.js");
var import_FormFieldMessageContext = require("../../contexts/FormFieldMessageContext.js");
var import_useFormFieldContext = require("../../hooks/useFormFieldContext.js");
var import_useFormFieldMessagesContext = require("../../hooks/useFormFieldMessagesContext.js");
var import_formfield_messages_state_reducer = require("../../state/formfield-messages-state-reducer.js");
const FormFieldMessageContextProvider = (formFieldProps) => {
  const generatedId = (0, import_useId.useId)();
  const {
    children,
    id: propId = generatedId,
    disabled: propDisabled
  } = formFieldProps;
  const formFieldState = (0, import_useFormFieldContext.useFormFieldContext)();
  const formFieldMessageReducerCB = (0, import_react.useCallback)(
    (state2, action) => (0, import_formfield_messages_state_reducer.formFieldMessagesReducer)(state2, action),
    []
  );
  const [state, dispatch] = (0, import_react.useReducer)(formFieldMessageReducerCB, {
    customMessages: [],
    formControlMessages: [],
    showError: false,
    isTouched: false,
    isValid: void 0,
    controlIds: [formFieldState.controlId ?? propId]
  });
  const formFieldMessageContextValue = (0, import_react.useMemo)(
    () => ({
      disabled: formFieldState.disabled || propDisabled,
      controlId: formFieldState.controlId || propId,
      dispatch,
      state
    }),
    [
      formFieldState.disabled,
      formFieldState.controlId,
      propDisabled,
      propId,
      dispatch,
      state
    ]
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_FormFieldMessageContext.FormFieldMessageContext.Provider, { value: formFieldMessageContextValue, children });
};
function withFormFieldMessageContextProviderWithGenerics(OriginalComponent) {
  const WrappedComponent = (0, import_react.forwardRef)((props, forwardedRef) => {
    const { dispatch } = (0, import_useFormFieldMessagesContext.useFormFieldMessagesContext)();
    if (dispatch !== void 0) {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(OriginalComponent, { ref: forwardedRef, ...props });
    }
    const { disabled, id } = props;
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(FormFieldMessageContextProvider, { disabled, id, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(OriginalComponent, { ref: forwardedRef, ...props }) });
  });
  return WrappedComponent;
}
function withFormFieldMessageContextProvider(OriginalComponent) {
  const WithFormFieldMessageContextProvider = (0, import_react.forwardRef)(
    (props, forwardedRef) => {
      const { dispatch } = (0, import_useFormFieldMessagesContext.useFormFieldMessagesContext)();
      if (dispatch !== void 0) {
        return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(OriginalComponent, { ref: forwardedRef, ...props });
      }
      const { disabled, id } = props;
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(FormFieldMessageContextProvider, { disabled, id, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(OriginalComponent, { ref: forwardedRef, ...props }) });
    }
  );
  return WithFormFieldMessageContextProvider;
}
FormFieldMessageContextProvider.displayName = "FormFieldMessageContextProvider";
