// @dynatrace/strato-components/forms/forms-core/components/form-field/FormFieldContextProvider.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var FormFieldContextProvider_exports = {};
__export(FormFieldContextProvider_exports, {
  FormFieldContextProvider: () => FormFieldContextProvider
});
module.exports = __toCommonJS(FormFieldContextProvider_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_useId = require("../../../../core/hooks/useId.js");
var import_parse_boolean = require("../../../../core/utils/parse-boolean.js");
var import_FormFieldContext = require("../../contexts/FormFieldContext.js");
var import_useFormFieldContext = require("../../hooks/useFormFieldContext.js");
var import_formfield_state_reducer = require("../../state/formfield-state-reducer.js");
const FormFieldContextProvider = (props) => {
  const generatedId = (0, import_useId.useId)();
  const {
    children,
    controlId: propId = generatedId,
    disabled: propDisabled,
    required: propRequired,
    ariaDisabled: ariaDisabledProp,
    readOnly: propReadOnly
  } = props;
  const {
    disabled: contextDisabled,
    ariaDisabled: contextAriaDisabled,
    readOnly: contextReadOnly
  } = (0, import_useFormFieldContext.useFormFieldContext)();
  const [formFieldState, formFieldDispatch] = (0, import_react.useReducer)(import_formfield_state_reducer._formFieldReducer, {});
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_FormFieldContext._FormFieldContext.Provider,
    {
      value: {
        controlId: formFieldState.controlId || propId,
        disabled: formFieldState.disabled ?? propDisabled ?? contextDisabled,
        required: formFieldState.required ?? propRequired,
        readOnly: formFieldState.readOnly ?? propReadOnly ?? contextReadOnly,
        dispatch: formFieldDispatch,
        ariaDisabled: formFieldState.ariaDisabled ?? (0, import_parse_boolean.parseBoolean)(ariaDisabledProp) ?? contextAriaDisabled
      },
      children
    }
  );
};
FormFieldContextProvider.displayName = "FormFieldContextProvider";
