// @dynatrace/strato-components/forms/forms-core/components/form-field/FormField.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var FormField_exports = {};
__export(FormField_exports, {
  FormField: () => FormField
});
module.exports = __toCommonJS(FormField_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var styles = __toESM(require("./FormField.sty.js"));
var import_FormFieldContextProvider = require("./FormFieldContextProvider.js");
var import_FormFieldMessageContextProvider = require("./FormFieldMessageContextProvider.js");
var import_useBehavioralTrackingProps = require("../../../../core/hooks/useBehavioralTrackingProps.js");
var import_useId = require("../../../../core/hooks/useId.js");
var import_merge_props = require("../../../../core/utils/merge-props.js");
var import_parse_boolean = require("../../../../core/utils/parse-boolean.js");
var import_Flex = require("../../../../layouts/flex/Flex.js");
const COMPONENT_NAME = "FormField";
const FormField = /* @__PURE__ */ (0, import_react.forwardRef)((formFieldProps, forwardedRef) => {
  const generatedId = (0, import_useId.useId)();
  const {
    children,
    id: propId = generatedId,
    disabled,
    required,
    readOnly,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "aria-disabled": ariaDisabledProp,
    ...remainingProps
  } = formFieldProps;
  const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_FormFieldContextProvider.FormFieldContextProvider,
    {
      controlId: propId,
      disabled,
      ariaDisabled: (0, import_parse_boolean.parseBoolean)(ariaDisabledProp),
      required,
      readOnly,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_FormFieldMessageContextProvider.FormFieldMessageContextProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_Flex.Flex,
        {
          ref: forwardedRef,
          "data-testid": dataTestId,
          "data-dtrum-mask": dataDtrumMask,
          "data-dtrum-allow": dataDtrumAllow,
          className: (0, import_clsx.default)(styles.formField, consumerClassName),
          gap: 4,
          flexDirection: "column",
          style: consumerStyle,
          ...(0, import_merge_props.mergeProps)(restProps, behavioralTrackingProps),
          "data-dt-component": COMPONENT_NAME,
          children
        }
      ) })
    }
  );
});
FormField.displayName = COMPONENT_NAME;
