// @dynatrace/strato-components/esm/forms/spin-buttons/spin-button-group/hooks/useNavigateSpinButton.js
import {
  useCallback
} from "react";
function handleArrowLeft(event, prevSpinButton, currentSpinButton, selection, onLeadingEdge) {
  if (prevSpinButton && selection?.start === 0) {
    prevSpinButton.element.current?.focus();
    event.preventDefault();
  } else if (!prevSpinButton && onLeadingEdge && selection?.start === 0) {
    onLeadingEdge();
    event.preventDefault();
  } else if (currentSpinButton && selection === void 0) {
    currentSpinButton.selectAll();
    event.preventDefault();
  } else if (!prevSpinButton && selection?.start === 0) {
    event.preventDefault();
  }
}
function handleArrowRight(event, currentSpinButton, nextSpinButton, selection, length, onTrailingEdge) {
  if (nextSpinButton && selection && selection.end === length) {
    nextSpinButton.element.current?.focus();
    event.preventDefault();
  } else if (!nextSpinButton && onTrailingEdge && selection && selection.end === length) {
    onTrailingEdge();
    event.preventDefault();
  } else if (currentSpinButton && selection === void 0) {
    currentSpinButton.selectAll();
    event.preventDefault();
  } else if (!nextSpinButton && selection?.end === length) {
    event.preventDefault();
  }
}
function handleDeletion(event, nextSpinButton, selection, length, onEdgeHandler) {
  if (!length && selection?.start === 0 && selection?.end === 0) {
    if (nextSpinButton) {
      nextSpinButton.element.current?.focus();
      event.preventDefault();
    } else if (onEdgeHandler) {
      event.preventDefault();
      requestAnimationFrame(() => onEdgeHandler());
    }
  }
}
function handleTriggerNavigation(e, spinButtonRefs, triggerRefs, target) {
  const lookupIndex = spinButtonRefs.current?.findIndex((h) => h === target);
  const nextTriggerIndex = triggerRefs.current.findIndex(
    (trigger2, index) => trigger2 !== void 0 && index > lookupIndex
  );
  const afterTriggerSpinButton = spinButtonRefs.current?.find(
    (spinbutton, index) => spinbutton && index > nextTriggerIndex
  );
  const trigger = triggerRefs.current[nextTriggerIndex];
  if (trigger !== void 0 && afterTriggerSpinButton && e.key === trigger) {
    afterTriggerSpinButton.element.current?.focus();
    e.preventDefault();
  }
}
function handleTrailingTextAdvance(e, target, next, onTrailingEdge) {
  if (target.type !== "text" || next || !/^[a-zA-Z]$/.test(e.key) || e.metaKey || e.ctrlKey) {
    return;
  }
  const match = target.searchValue(e.key);
  if (match !== void 0) {
    onTrailingEdge?.();
  }
}
const useNavigateSpinButton = (spinButtonRefs, triggerRefs, onLeadingEdge, onTrailingEdge) => {
  const onKeyDownHandler = useCallback(
    (e) => {
      if (e.key === "Shift") {
        return;
      }
      const handlers = spinButtonRefs.current.filter(Boolean);
      const index = handlers?.findIndex(
        ({ element }) => element?.current === e?.target
      );
      if (index < 0) {
        return;
      }
      const [prev, target, next] = [
        handlers[index - 1],
        handlers[index],
        handlers[index + 1]
      ];
      const length = target?.element?.current?.textContent?.length;
      const selection = target?.getSelection();
      if (!e.shiftKey) {
        switch (e.key) {
          case "ArrowLeft":
            handleArrowLeft(e, prev, target, selection, onLeadingEdge);
            return;
          case "ArrowRight":
            handleArrowRight(
              e,
              target,
              next,
              selection,
              length,
              onTrailingEdge
            );
            return;
          case "Backspace":
            handleDeletion(e, prev, selection, length, onLeadingEdge);
            return;
          case "Delete":
            handleDeletion(e, next, selection, length, onTrailingEdge);
            return;
        }
      }
      handleTriggerNavigation(e, spinButtonRefs, triggerRefs, target);
      handleTrailingTextAdvance(e, target, next, onTrailingEdge);
    },
    [spinButtonRefs, triggerRefs, onLeadingEdge, onTrailingEdge]
  );
  return { onKeyDownHandler };
};
export {
  useNavigateSpinButton
};
