// @dynatrace/strato-components/esm/charts/meter-bar/shared/components/meter-bar-layout/components/MeterBarLayoutTopRightSlot.js
import { jsx, jsxs } from "react/jsx-runtime";
import {
  forwardRef,
  useImperativeHandle,
  useRef,
  useState
} from "react";
import { MeterBarChartToolbar } from "../../meter-bar-toolbar/MeterBarToolbar.js";
import { useComputedSegmentSize } from "../hooks/useComputedSegmentSize.js";
import {
  toolbarAnimatedContainer,
  valueAnimatedContainer,
  valueContainer
} from "../MeterBarLayout.sty.js";
const toolbarWidth = "24px";
const toolbarMarginLeft = "4px";
const MeterBarLayoutTopRightSlot = forwardRef(({ isToolbarEnabled, value }, forwardedRef) => {
  const [isChartHovered, setIsChartHovered] = useState(false);
  const popoverRef = useRef(null);
  const { valueLabelRef } = useComputedSegmentSize();
  const showPopover = () => setIsChartHovered(true);
  const hidePopover = () => setIsChartHovered(false);
  useImperativeHandle(forwardedRef, () => ({
    showPopover,
    hidePopover
  }));
  const isToolbarShown = isChartHovered && isToolbarEnabled;
  return /* @__PURE__ */ jsxs(
    "div",
    {
      ref: valueLabelRef,
      className: valueContainer({ inline: isToolbarEnabled }),
      children: [
        value && /* @__PURE__ */ jsx("div", { className: valueAnimatedContainer, children: value }),
        isToolbarEnabled && /* @__PURE__ */ jsx(
          "div",
          {
            ref: popoverRef,
            className: toolbarAnimatedContainer,
            style: {
              width: isToolbarShown ? toolbarWidth : 0,
              opacity: isToolbarShown ? 1 : 0,
              marginLeft: isToolbarShown ? toolbarMarginLeft : 0
            },
            children: /* @__PURE__ */ jsx(MeterBarChartToolbar, {})
          }
        )
      ]
    }
  );
});
MeterBarLayoutTopRightSlot.displayName = "MeterBarLayoutTopRightSlot";
export {
  MeterBarLayoutTopRightSlot
};
