// @dynatrace/strato-geo/map/components/DotLayer/DotLayer.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DotLayer_exports = {};
__export(DotLayer_exports, {
  DotLayer: () => DotLayer
});
module.exports = __toCommonJS(DotLayer_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_maplibre = require("@vis.gl/react-maplibre");
var import_core = require("@dynatrace/strato-components/core");
var import_BackgroundLayer = require("./BackgroundLayer.js");
var import_BackgroundSymbolLayer = require("./BackgroundSymbolLayer.js");
var import_DotLayerTooltip = require("./DotLayerTooltip.js");
var import_parse_dot_data_to_geo_json = require("./utils/parse-dot-data-to-geo-json.js");
var import_constants = require("../../constants.js");
var import_use_layer_coloring_strategy = require("../../hooks/use-layer-coloring-strategy.js");
var import_use_legend_interaction_highlighting = require("../../hooks/use-legend-interaction-highlighting.js");
var import_use_legend_interaction_visibility = require("../../hooks/use-legend-interaction-visibility.js");
var import_use_load_symbol_into_map = require("../../hooks/use-load-symbol-into-map.js");
var import_use_resolve_color = require("../../hooks/use-resolve-color.js");
var import_use_tooltip_template = require("../../hooks/use-tooltip-template.js");
var import_DotLayerTooltip2 = require("../../slots/DotLayerTooltip.js");
var import_tooltip_type_guards = require("../../utils/tooltip-type-guards.js");
var import_TooltipWrapper = require("../TooltipWrapper.js");
const DotLayer = (props) => {
  const [behavioralTrackingProps] = (0, import_core._useBehavioralTrackingProps)(props);
  const {
    data,
    layerId,
    bearing = 0,
    children,
    shape = "pin",
    background = false,
    shapeSize = import_constants.DEFAULT_SYMBOL_SIZE
  } = props;
  const calculatedShapeSize = shapeSize > 0 ? shapeSize : import_constants.FALLBACK_SYMBOL_SIZE;
  const { symbolName, symbolType } = (0, import_use_load_symbol_into_map.useLoadSymbolIntoMap)(
    shape,
    layerId,
    calculatedShapeSize
  );
  const bgLayerId = `${layerId}-background`;
  const { symbolName: bgName } = (0, import_use_load_symbol_into_map.useLoadSymbolIntoMap)(
    "circle",
    bgLayerId,
    calculatedShapeSize
  );
  const defaultSymbolColor = symbolType === "shape" ? import_constants.DEFAULT_SHAPE_COLOR : import_constants.DEFAULT_ICON_COLOR;
  const colorParser = (0, import_use_layer_coloring_strategy.useLayerColoringStrategy)();
  const shapeColor = (0, import_use_resolve_color.useResolveLocationColor)(
    defaultSymbolColor,
    props,
    colorParser
  );
  const {
    handler: tooltipTemplate,
    hidden: tooltipHidden,
    symbolAlignment
  } = (0, import_use_tooltip_template.useTooltipTemplate)(children, import_DotLayerTooltip2.DotLayerTooltip, { extractProps: true });
  const sourceId = `source-${layerId}`;
  (0, import_use_legend_interaction_highlighting.useLegendInteractionHighlighting)(props, sourceId);
  (0, import_use_legend_interaction_visibility.useLegendInteractionVisibility)(props, layerId);
  const dataGeoJson = (0, import_parse_dot_data_to_geo_json.parseDotDataToGeoJSON)(
    data,
    shapeColor,
    bearing,
    background
  );
  const iconAnchor = shape === "pin" ? "bottom" : "center";
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_react_maplibre.Source, { id: sourceId, type: "geojson", data: dataGeoJson, generateId: true, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_react_maplibre.Layer,
      {
        type: "symbol",
        id: layerId,
        layout: {
          "icon-image": symbolName,
          "icon-size": import_constants.DEFAULT_ICON_SIZE_RATIO,
          "icon-rotate": ["get", "__bearing"],
          // https://maplibre.org/maplibre-style-spec/layers/#layout-symbol-icon-ignore-placement
          "icon-ignore-placement": true,
          "icon-anchor": iconAnchor
        },
        paint: {
          "icon-color": ["get", "__color"],
          "icon-opacity": [
            "case",
            ["boolean", ["feature-state", "legendHover"], true],
            import_constants.SHAPE_OPACITY_DEFAULT,
            import_constants.SHAPE_OPACITY_DIMMED
          ]
        },
        ...behavioralTrackingProps
      }
    ),
    shape !== "pin" ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_BackgroundLayer.BackgroundLayer,
      {
        background: Boolean(background),
        beforeId: layerId,
        source: sourceId,
        radius: calculatedShapeSize / 2,
        iconAnchor
      }
    ) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_BackgroundSymbolLayer.BackgroundSymbolLayer,
      {
        background: Boolean(background),
        beforeId: layerId,
        source: sourceId,
        radius: calculatedShapeSize / 2,
        iconAnchor,
        symbolName: bgName,
        layerId: bgLayerId,
        shape
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TooltipWrapper.TooltipWrapper, { layerId, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_DotLayerTooltip.DotLayerTooltip,
      {
        tooltipTemplate: (0, import_tooltip_type_guards.isDotTooltipTemplate)(tooltipTemplate) ? tooltipTemplate : void 0,
        hidden: tooltipHidden,
        symbolAlignment
      }
    ) })
  ] });
};
DotLayer["displayName"] = "DotLayer";
