// @dynatrace/strato-components/esm/tables/DataTable/components/Toolbar/Download/DataTableDownloadData.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { useIntl } from "react-intl";
import * as styles from "./DataTableDownloadData.sty.js";
import { Menu } from "../../../../../navigation/menu/Menu.js";
import { useDataTable } from "../../../contexts/DataTableContext.js";
import { dataTableMessages } from "../../../messages.js";
const DataTableDownloadData = ({
  excludeColumns,
  hasIntents,
  isDownloading,
  onDownloadStart,
  onDownloadEnd
}) => {
  const intl = useIntl();
  const table = useDataTable();
  const selectionCount = table.getSelectionCount();
  const menuGroupLabel = intl.formatMessage(dataTableMessages.downloadTable);
  const downloadMenuItems = /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsxs(
      Menu.Item,
      {
        onSelect: () => {
          if (isDownloading) {
            return;
          }
          onDownloadStart();
          table.downloadAllData(excludeColumns, onDownloadEnd);
        },
        className: styles.downloadMenuItem,
        "data-dt-name": "ToolbarDownloadAll",
        children: [
          intl.formatMessage(dataTableMessages.downloadAll),
          /* @__PURE__ */ jsx(Menu.Suffix, { className: styles.downloadMenuSuffix, children: intl.formatMessage(dataTableMessages.downloadRowsCount, {
            noOfRows: table.getPrePaginationRowModel().flatRows.length
          }) })
        ]
      }
    ),
    table.options.enablePagination && /* @__PURE__ */ jsxs(
      Menu.Item,
      {
        onSelect: () => {
          if (isDownloading) {
            return;
          }
          onDownloadStart();
          table.downloadPageData(excludeColumns, onDownloadEnd);
        },
        className: styles.downloadMenuItem,
        "data-dt-name": "ToolbarDownloadPage",
        children: [
          intl.formatMessage(dataTableMessages.downloadPage),
          /* @__PURE__ */ jsx(Menu.Suffix, { className: styles.downloadMenuSuffix, children: intl.formatMessage(dataTableMessages.downloadRowsCount, {
            noOfRows: table.getPaginationRowModel().flatRows.length
          }) })
        ]
      }
    ),
    table.options.enableRowSelection && /* @__PURE__ */ jsxs(
      Menu.Item,
      {
        disabled: selectionCount === 0,
        onSelect: () => {
          if (isDownloading) {
            return;
          }
          onDownloadStart();
          table.downloadSelectedData(excludeColumns, onDownloadEnd);
        },
        className: styles.downloadMenuItem,
        "data-dt-name": "ToolbarDownloadSelected",
        children: [
          intl.formatMessage(dataTableMessages.downloadSelected),
          /* @__PURE__ */ jsx(Menu.Suffix, { className: styles.downloadMenuSuffix, children: intl.formatMessage(dataTableMessages.downloadRowsCount, {
            noOfRows: selectionCount
          }) })
        ]
      }
    )
  ] });
  return hasIntents ? /* @__PURE__ */ jsxs(Menu.Sub, { defaultOpen: false, children: [
    /* @__PURE__ */ jsx(Menu.SubTrigger, { children: menuGroupLabel }),
    /* @__PURE__ */ jsx(Menu.SubContent, { children: downloadMenuItems })
  ] }) : /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(Menu.Label, { children: menuGroupLabel }),
    downloadMenuItems
  ] });
};
export {
  DataTableDownloadData
};
