// @dynatrace/strato-components/forms/select/hooks/use-select-slots.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var use_select_slots_exports = {};
__export(use_select_slots_exports, {
  useSelectSlots: () => useSelectSlots
});
module.exports = __toCommonJS(use_select_slots_exports);
var import_react = require("react");
var import_is_slot = require("../../../core/utils/is-slot.js");
var import_SelectContent = require("../SelectContent.js");
var import_SelectCustomTrigger = require("../SelectCustomTrigger.js");
var import_SelectDisplayValue = require("../SelectDisplayValue.js");
var import_SelectEmptyState = require("../SelectEmptyState.js");
var import_SelectFilter = require("../SelectFilter.js");
var import_SelectPrefix = require("../SelectPrefix.js");
var import_SelectTrigger = require("../SelectTrigger.js");
function useSelectSlots(children) {
  return (0, import_react.useMemo)(() => {
    const slots = {
      Content: null,
      CustomTrigger: null,
      DisplayValue: null,
      Trigger: null,
      Prefix: null,
      Filter: null,
      EmptyState: null
    };
    import_react.Children.forEach(children, (child) => {
      if ((0, import_is_slot.isSlot)(child, import_SelectContent.SelectContent)) {
        slots.Content = child;
        import_react.Children.forEach(child.props.children, (child2) => {
          if ((0, import_is_slot.isSlot)(child2, import_SelectFilter.SelectFilter)) {
            slots.Filter = child2;
            return;
          } else if ((0, import_is_slot.isSlot)(child2, import_SelectEmptyState.SelectEmptyState)) {
            slots.EmptyState = child2;
          }
        });
        return;
      } else if ((0, import_is_slot.isSlot)(child, import_SelectTrigger.SelectTrigger)) {
        slots.Trigger = child;
        import_react.Children.forEach(child.props.children, (child2) => {
          if ((0, import_is_slot.isSlot)(child2, import_SelectDisplayValue.SelectDisplayValue)) {
            import_react.Children.forEach(child2.props.children, (child3) => {
              if ((0, import_is_slot.isSlot)(child3, import_SelectPrefix.SelectPrefix)) {
                slots.Prefix = child3;
                return;
              }
              slots.DisplayValue = slots.DisplayValue || [];
              slots.DisplayValue.push(child3);
            });
          }
        });
      } else if ((0, import_is_slot.isSlot)(child, import_SelectCustomTrigger.SelectCustomTrigger)) {
        const triggerValue = import_react.Children.toArray(child.props.children)[0];
        if ((0, import_react.isValidElement)(triggerValue)) {
          slots.CustomTrigger = triggerValue;
        }
      } else if ((0, import_is_slot.isSlot)(child, import_SelectFilter.SelectFilter)) {
        slots.Filter = child;
      }
    });
    return slots;
  }, [children]);
}
