// @dynatrace/strato-components/forms/select/hooks/use-select-options.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var use_select_options_exports = {};
__export(use_select_options_exports, {
  _isGroup: () => _isGroup,
  _isOption: () => _isOption,
  safeStringify: () => safeStringify,
  useSelectOptions: () => useSelectOptions
});
module.exports = __toCommonJS(use_select_options_exports);
var import_react = require("react");
var import_can_receive_default_text_format = require("../../../core/utils/can-receive-default-text-format.js");
var import_extract_behavior_tracking_props = require("../../../core/utils/extract-behavior-tracking-props.js");
var import_is_slot = require("../../../core/utils/is-slot.js");
var import_SelectGroup = require("../SelectGroup.js");
var import_SelectOption = require("../SelectOption.js");
var import_SelectPrefix = require("../SelectPrefix.js");
var import_SelectSuffix = require("../SelectSuffix.js");
function _isOption(optionOrGroup) {
  if (optionOrGroup.value !== void 0) {
    return true;
  }
  return false;
}
function _isGroup(optionOrGroup) {
  if (Array.isArray(optionOrGroup.options)) {
    return true;
  }
  return false;
}
function extractOptionData(optionElement) {
  const optionSlots = {
    prefix: { Prefix: null },
    Label: null,
    suffix: { Suffix: null }
  };
  import_react.Children.forEach(optionElement.props.children, (child) => {
    if ((0, import_is_slot.isSlot)(child, import_SelectPrefix.SelectPrefix)) {
      optionSlots.prefix.Prefix = child.props.children;
      optionSlots.prefix.componentProps = {
        className: child.props.className,
        style: child.props.style,
        "data-testid": child.props["data-testid"],
        ...(0, import_extract_behavior_tracking_props.extractBehaviorTrackingProps)(child.props)
      };
    } else if ((0, import_is_slot.isSlot)(child, import_SelectSuffix.SelectSuffix)) {
      optionSlots.suffix.Suffix = child.props.children;
      optionSlots.suffix.componentProps = {
        className: child.props.className,
        style: child.props.style,
        "data-testid": child.props["data-testid"],
        ...(0, import_extract_behavior_tracking_props.extractBehaviorTrackingProps)(child.props)
      };
    } else {
      optionSlots.Label = optionSlots.Label ?? [];
      if (child !== null) {
        optionSlots.Label.push(child);
      }
    }
  });
  return {
    value: optionElement.props.value,
    prefix: {
      component: optionSlots.prefix.Prefix,
      componentProps: optionSlots.prefix.componentProps
    },
    label: optionSlots.Label,
    suffix: {
      component: optionSlots.suffix.Suffix,
      componentProps: optionSlots.suffix.componentProps
    },
    isDisabled: optionElement.props.disabled,
    textValue: optionElement.props.textValue,
    "aria-label": optionElement.props["aria-label"],
    className: optionElement.props.className,
    style: optionElement.props.style,
    "data-testid": optionElement.props["data-testid"],
    ...(0, import_extract_behavior_tracking_props.extractBehaviorTrackingProps)(optionElement.props)
  };
}
function safeStringify(value) {
  try {
    if (typeof value === "string") {
      return value;
    }
    return JSON.stringify(value);
  } catch {
    return "";
  }
}
function reportNoStringifyableValueError(value, logger) {
  logger?.warn("Failed to stringify value.", value);
}
function getOptionLabel(option, fallbackLabel) {
  if (option.textValue) {
    return option.textValue;
  }
  const labelIsStringifyable = (0, import_can_receive_default_text_format.canReceiveDefaultTextFormat)(option.label);
  const stringifiedValue = safeStringify(option.value);
  if (labelIsStringifyable) {
    return `${option.label}`;
  }
  if (stringifiedValue) {
    return stringifiedValue;
  }
  return fallbackLabel;
}
function getOptionValue(option, fallbackValue, logger) {
  if (option.textValue) {
    return option.textValue;
  }
  const stringifiedValue = safeStringify(option.value);
  if (stringifiedValue) {
    return stringifiedValue;
  }
  const labelIsStringifyable = (0, import_can_receive_default_text_format.canReceiveDefaultTextFormat)(option.label);
  if (labelIsStringifyable) {
    return `${option.label}`;
  }
  reportNoStringifyableValueError(option.value, logger);
  return fallbackValue;
}
function extractGroupData(groupElement, groupIndex, logger) {
  const {
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId
  } = groupElement.props;
  const group = {
    label: void 0,
    options: [],
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    ...(0, import_extract_behavior_tracking_props.extractBehaviorTrackingProps)(groupElement.props)
  };
  import_react.Children.forEach(groupElement?.props.children, (child, optionIndex) => {
    if ((0, import_is_slot.isSlot)(child, import_SelectGroup.SelectGroupLabel)) {
      group.label = child.props.children;
      group.labelProps = {
        className: child.props.className,
        style: child.props.style,
        "data-testid": child.props["data-testid"],
        ...(0, import_extract_behavior_tracking_props.extractBehaviorTrackingProps)(child.props)
      };
    }
    if ((0, import_SelectOption._isSelectOptionElement)(child)) {
      const option = extractOptionData(child);
      if (option.label) {
        const stringifiedLabel = getOptionLabel(
          option,
          `group-${groupIndex}-label-${optionIndex}`
        );
        option.label.toString = () => stringifiedLabel;
      }
      if (!option.textValue) {
        const stringifiedValue = getOptionValue(
          option,
          `group-${groupIndex}-label-${optionIndex}`,
          logger
        );
        option.textValue = stringifiedValue;
      }
      if (option !== null) {
        group.options.push(option);
      }
    }
  });
  return group;
}
function useSelectOptions(contentSlot, logger) {
  return (0, import_react.useMemo)(() => {
    const optionsOrGroups = [];
    let optionsHavePrefix = false;
    const allOptions = [];
    import_react.Children.forEach(contentSlot?.props.children, (child, childIndex) => {
      if ((0, import_SelectOption._isSelectOptionElement)(child)) {
        const option = extractOptionData(child);
        if (option.label) {
          const stringifiedLabel = getOptionLabel(
            option,
            `label-${childIndex}`
          );
          option.label.toString = () => stringifiedLabel;
        }
        if (!option.textValue) {
          const stringifiedValue = getOptionValue(
            option,
            `value-${childIndex}`,
            logger
          );
          option.textValue = stringifiedValue;
        }
        if (option !== null) {
          optionsOrGroups.push(option);
          if (option.prefix.component !== null) {
            optionsHavePrefix = true;
          }
        }
        allOptions.push(option);
      }
      if ((0, import_is_slot.isSlot)(child, import_SelectGroup.SelectGroup)) {
        const group = extractGroupData(child, childIndex, logger);
        if (group !== null) {
          optionsOrGroups.push(group);
        }
        if (group.options.length > 2e4) {
          logger?.warn(
            `Group with label "${group.label}" has more than 20,000 options. This can lead to a slow and unresponsive UI.`
          );
        }
        for (const element of group.options) {
          allOptions.push(element);
        }
        if (optionsHavePrefix === false) {
          const nestedOptionsHavePrefix = group.options.some(
            (option) => option.prefix.component !== null
          );
          if (nestedOptionsHavePrefix) {
            optionsHavePrefix = true;
          }
        }
      }
    });
    return { optionsOrGroups, optionsHavePrefix, allOptions };
  }, [contentSlot, logger]);
}
