// @dynatrace/strato-components/charts/core/components/chart-layout/components/ResizableChartLayout.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ResizableChartLayout_exports = {};
__export(ResizableChartLayout_exports, {
  ResizableChartLayout: () => ResizableChartLayout
});
module.exports = __toCommonJS(ResizableChartLayout_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_styles_css = require("./styles.sty.js");
var import_Flex = require("../../../../../layouts/flex/Flex.js");
var import_LegendSizeContext = require("../../legend/shared/context/LegendSizeContext.js");
var import_GraphSize_context = require("../context/GraphSize.context.js");
var import_useChartLayoutLoading = require("../hooks/useChartLayoutLoading.js");
var import_useChartLayoutOrientation = require("../hooks/useChartLayoutOrientation.js");
var import_useChartLayoutSize = require("../hooks/useChartLayoutSize.js");
var import_useChartLayoutTemplate = require("../hooks/useChartLayoutTemplate.js");
var import_useResizableLayout = require("../hooks/useResizableLayout.js");
var import_orientation = require("../utils/orientation.js");
const ResizableChartLayout = ({
  onResize,
  showResizer,
  showLegend
}) => {
  const { orientation } = (0, import_useChartLayoutOrientation.useChartLayoutOrientation)();
  const loading = (0, import_useChartLayoutLoading.useChartLayoutLoading)();
  const containerUsableArea = (0, import_useChartLayoutSize.useChartLayoutSize)();
  const { legendArea, graphArea } = (0, import_useChartLayoutTemplate.useChartLayoutTemplate)();
  const isVertical = (0, import_orientation.isVertical)(orientation);
  const {
    legendWidth,
    legendHeight,
    graphSize,
    legendContainerRef,
    legendResizeLineRef,
    legendResizeContainerRef
  } = (0, import_useResizableLayout.useResizableLayout)(
    containerUsableArea.width,
    containerUsableArea.height,
    orientation,
    showResizer,
    showLegend,
    onResize
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: (0, import_styles_css.graphContainer)({ loading }), children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_GraphSize_context.GraphSizeContext.Provider, { value: graphSize, children: graphArea }) }),
    showLegend && showResizer && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Flex.Flex,
      {
        ref: legendResizeContainerRef,
        alignItems: "center",
        justifyContent: isVertical ? "center" : "flex-end",
        className: (0, import_styles_css.resizeContainer)({ orientation, loading }),
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          "div",
          {
            "data-testid": "legend-resize-line",
            ref: legendResizeLineRef,
            className: (0, import_styles_css.legendResizeLine)({ orientation })
          }
        )
      }
    ),
    showLegend && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "div",
      {
        ref: legendContainerRef,
        className: (0, import_styles_css.legendContainer)({
          orientation,
          loading
        }),
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_LegendSizeContext.LegendSizeContext.Provider,
          {
            value: { width: legendWidth, height: legendHeight },
            children: legendArea
          }
        )
      }
    )
  ] });
};
ResizableChartLayout["displayName"] = "ResizableChartLayout";
