// @dynatrace/strato-components/esm/content/ai-response/tests/vite-helpers.js
import { jsx } from "react/jsx-runtime";
import {
  render as rtlRender
} from "@testing-library/react";
import { IntlProvider } from "react-intl";
import { vi } from "vitest";
const pristineMatchMedia = window.matchMedia;
function setupMatchMediaMock(isMatching = false) {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: vi.fn().mockImplementation((query) => ({
      matches: isMatching,
      media: query,
      onchange: null,
      addListener: vi.fn(),
      // Deprecated
      removeListener: vi.fn(),
      // Deprecated
      addEventListener: vi.fn(),
      removeEventListener: vi.fn(),
      dispatchEvent: vi.fn()
    }))
  });
}
function clearMatchMediaMock() {
  window.matchMedia = pristineMatchMedia;
}
const AllProviders = ({
  children,
  locale = "en",
  timeZone = "UTC"
}) => {
  return /* @__PURE__ */ jsx(IntlProvider, { locale, timeZone, children });
};
function render(ui, {
  locale = "en",
  timeZone = "UTC",
  ...renderOptions
} = {}) {
  return rtlRender(ui, {
    wrapper: ({ children }) => /* @__PURE__ */ jsx(AllProviders, { locale, timeZone, children }),
    ...renderOptions
  });
}
export {
  clearMatchMediaMock,
  render,
  setupMatchMediaMock
};
