// @dynatrace/strato-components/forms/forms-core/time/date-time.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var date_time_exports = {};
__export(date_time_exports, {
  GRAIL_MAX_DATE: () => GRAIL_MAX_DATE,
  GRAIL_MIN_DATE: () => GRAIL_MIN_DATE,
  applyInitialTime: () => applyInitialTime,
  getFormattedDateTime: () => getFormattedDateTime,
  getSecondFractions: () => getSecondFractions,
  replaceSecondFractions: () => replaceSecondFractions
});
module.exports = __toCommonJS(date_time_exports);
var import_tz = require("@date-fns/tz");
var import_date_fns = require("date-fns");
var import_user_preferences = require("@dynatrace-sdk/user-preferences");
const GRAIL_MIN_DATE = "1677-09-21T00:12:43.145224192Z";
const GRAIL_MAX_DATE = "2262-04-11T23:47:16.854775807Z";
function getISOFormattedDate(date) {
  return `${valueTo4DigitString(date.getFullYear())}-${valueTo2DigitString(
    date.getMonth() + 1
  )}-${valueTo2DigitString(date.getDate())}`;
}
function getISOFormattedTime(date) {
  return `${valueTo2DigitString(date.getHours())}:${valueTo2DigitString(
    date.getMinutes()
  )}:${valueTo2DigitString(date.getSeconds())}`;
}
function getFormattedDateTime(date) {
  return `${getISOFormattedDate(date)} ${getISOFormattedTime(date)}`;
}
function getSecondFractions(searchString) {
  return searchString && searchString.includes(".") ? searchString.split(".")[1].replace("Z", "") : "";
}
function valueTo2DigitString(value) {
  return `${value}`.padStart(2, "0");
}
function valueTo4DigitString(value) {
  return `${value}`.padStart(4, "0");
}
function replaceSecondFractions(searchString, milliMicroNanoSeconds) {
  if (milliMicroNanoSeconds) {
    return `${searchString.replace(
      searchString.split(".")[1],
      milliMicroNanoSeconds
    )}Z`;
  }
  return searchString;
}
function getTimeFromISOString(isoDate) {
  return isoDate ? isoDate.split("T")[1].replace("Z", "") : "";
}
function applyInitialTime(newDate, originalDate, milliMicroNanoSeconds = getTimeFromISOString(
  (0, import_date_fns.startOfDay)(import_tz.TZDateMini.tz((0, import_user_preferences.getTimezone)())).toISOString()
).split(".")[1]) {
  if (!originalDate) {
    return replaceSecondFractions(newDate, milliMicroNanoSeconds);
  }
  const configuredTimezone = (0, import_user_preferences.getTimezone)();
  const datePart = new import_tz.TZDateMini(newDate, configuredTimezone);
  const timePart = new import_tz.TZDateMini(originalDate, configuredTimezone);
  datePart.setMilliseconds(timePart.getMilliseconds());
  datePart.setUTCSeconds(timePart.getUTCSeconds());
  datePart.setMinutes(timePart.getMinutes());
  datePart.setHours(timePart.getHours());
  return datePart.toISOString();
}
