// @dynatrace/strato-components/forms/forms-core/time/helpers.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var helpers_exports = {};
__export(helpers_exports, {
  convertDateOnlyToDateTime: () => convertDateOnlyToDateTime,
  eachDayOfInterval: () => eachDayOfInterval,
  filterOutGrailLimit: () => filterOutGrailLimit,
  getDayPeriodValues: () => getDayPeriodValues,
  getSafeRegionalFormat: () => getSafeRegionalFormat,
  getTimeValueFromSimpleValue: () => getTimeValueFromSimpleValue,
  resetRegionalFormatErrorFlag: () => resetRegionalFormatErrorFlag,
  sortTimeframeValues: () => sortTimeframeValues,
  timeValueToDate: () => timeValueToDate,
  toCalendarValue: () => toCalendarValue
});
module.exports = __toCommonJS(helpers_exports);
var import_tz = require("@date-fns/tz");
var import_date_fns = require("date-fns");
var import_units = require("@dynatrace-sdk/units");
var import_user_preferences = require("@dynatrace-sdk/user-preferences");
var import_date_time = require("./date-time.js");
var import_timeParser = require("./timeParser.js");
var import_sdk_logger = require("../../../core/sdk/sdk-logger.js");
var import_helpers = require("../../../core/time/helpers.js");
var import_validation = require("../../../core/time/validation.js");
var import_timeframe_selection = require("../../../filters/timeframe-selector/utils/timeframe-selection.js");
function normalizeDates(context, ...dates) {
  const normalize = import_date_fns.constructFrom.bind(
    null,
    context || dates.find((date) => typeof date === "object")
  );
  return dates.map(normalize);
}
function eachDayOfInterval(interval, options) {
  const [start, end] = normalizeDates(
    options?.in,
    interval.start,
    interval.end
  );
  let reversed = +start > +end;
  const endTime = reversed ? +start : +end;
  const date = reversed ? end : start;
  date.setHours(0, 0, 0, 0);
  let step = options?.step ?? 1;
  if (!step) {
    return [];
  }
  if (step < 0) {
    step = -step;
    reversed = !reversed;
  }
  const dates = [];
  while (+date <= endTime) {
    dates.push((0, import_date_fns.constructFrom)(start, date));
    date.setHours(date.getHours() + step * 24 + 1);
    date.setHours(0);
  }
  return reversed ? dates.reverse() : dates;
}
function sortTimeframeValues(value, min, max, precision) {
  if (value?.from?.absoluteDate && value?.to?.absoluteDate && (0, import_date_fns.isBefore)(value.to.absoluteDate, value.from.absoluteDate)) {
    return {
      from: getTimeValueFromSimpleValue(
        value.to.value,
        value.from.value,
        "from",
        value.from.type === "expression",
        min,
        max,
        precision
      ),
      to: getTimeValueFromSimpleValue(
        value.from.value,
        value.to.value,
        "to",
        value.to.type === "expression",
        min,
        max,
        precision
      )
    };
  }
  return {
    from: value?.from,
    to: value?.to
  };
}
let hasLoggedSafeRegionalFormatError = false;
function resetRegionalFormatErrorFlag() {
  hasLoggedSafeRegionalFormatError = false;
}
function getSafeRegionalFormat() {
  const region = (0, import_user_preferences.getRegionalFormat)();
  const modifiedRegion = removeProblematicLanguagesFromRegion(region);
  try {
    new Intl.Locale(modifiedRegion);
    if (Intl.DisplayNames.supportedLocalesOf([modifiedRegion]).length === 0) {
      throw new Error(
        `Locale "${modifiedRegion}" is not supported by Intl.DisplayNames.`
      );
    }
    return modifiedRegion;
  } catch (error) {
    const logger = (0, import_sdk_logger.getStratoLogger)("Date-Component");
    if (!hasLoggedSafeRegionalFormatError) {
      logger.warn(`Invalid regional format "${region}" encountered.`, error, {
        region,
        modifiedRegion
      });
      hasLoggedSafeRegionalFormatError = true;
    }
    return "en-US";
  }
}
const removeProblematicLanguagesFromRegion = (region) => {
  const problematicLanguages = [
    // Non-Arabic numerals.
    "ar",
    "as",
    "bgc",
    "bho",
    "bn",
    "ccp",
    "ckb",
    "dz",
    "fa",
    "ff",
    "ks",
    "lrc",
    "mni",
    "mr",
    "my",
    "mzn",
    "ne",
    "pa",
    "ps",
    "raj",
    "sa",
    "sat",
    "sd",
    "ur",
    "uz",
    // Problematic day period.
    "ak",
    "bem",
    "bo",
    "ee",
    "ha",
    "hi",
    "kok",
    "mai",
    "naq",
    "nus",
    "pt",
    "sq",
    "ti",
    "to",
    "yue",
    "zh",
    "tw",
    "bh",
    // Non gregorian calendars.
    "th"
  ];
  let safeRegionalFormat = region;
  if (!safeRegionalFormat) {
    return "en-US";
  }
  if (safeRegionalFormat.split("-")[1] === void 0) {
    return "en-" + safeRegionalFormat.split("-")[0];
  }
  if (problematicLanguages.find(
    (lang) => lang === safeRegionalFormat.split("-")[0].toLocaleLowerCase()
  )) {
    safeRegionalFormat = "en-" + safeRegionalFormat.split("-")[1];
  }
  if (safeRegionalFormat.split("-")[3]) {
    return safeRegionalFormat.split("-").slice(0, 2).join("-");
  }
  return safeRegionalFormat;
};
const getDayPeriodValues = () => {
  const dayPeriod = new Intl.DisplayNames(getSafeRegionalFormat(), {
    type: "dateTimeField"
  }).of("dayPeriod");
  if (!dayPeriod?.split("/")[1]) {
    return ["AM", "PM"];
  }
  return dayPeriod.split("/");
};
function timeValueToDate(timeValue, relativeDate) {
  if (!timeValue) {
    return null;
  }
  const parsedString = (0, import_units.parseTimeAsTimeValue)(timeValue?.value, relativeDate);
  return parsedString?.absoluteDate ? new import_tz.TZDateMini(parsedString.absoluteDate, (0, import_user_preferences.getTimezone)()) : null;
}
const getTimeValueFromSimpleValue = (candidate, timeOffset, position, isExpression, min, max, precision) => {
  if (!candidate) {
    return null;
  }
  const stringDate = typeof candidate === "string" ? candidate : candidate.toISOString();
  const timezone = (0, import_user_preferences.getTimezone)();
  const date = new import_tz.TZDateMini(stringDate, timezone);
  const minDate = new import_tz.TZDateMini(
    (0, import_timeParser.removePrecision)(min?.absoluteDate ?? "", precision),
    timezone
  );
  if (min && position !== "to" && (0, import_validation.isOutOfBounds)((0, import_date_fns.startOfDay)(date), min) && (0, import_date_fns.isSameDay)(minDate, date)) {
    if (min.type === "expression") {
      return min;
    }
    const offsetByPrecision = { milliseconds: 0, seconds: 1, minutes: 60 };
    const offsetSeconds = (
      // Only add the offset when the min has different precision than the calendar.
      minDate.getTime() === new import_tz.TZDateMini(min.absoluteDate, timezone).getTime() ? 0 : offsetByPrecision[precision] ?? 0
    );
    const parsedMin = (0, import_units.parseTimeAsTimeValue)(
      (0, import_date_fns.addSeconds)(minDate, offsetSeconds).toISOString(),
      void 0,
      precision
    );
    if (parsedMin && (0, import_date_fns.isSameDay)(minDate, date)) {
      return parsedMin;
    }
  }
  if (max && position !== "from" && (0, import_validation.isOutOfBounds)((0, import_date_fns.endOfDay)(date), void 0, max) && (0, import_date_fns.isSameDay)(new import_tz.TZDateMini(max.absoluteDate, timezone), date)) {
    return max;
  }
  const parsedDate = (0, import_units.parseTimeAsTimeValue)(stringDate, void 0, precision);
  if (!parsedDate) {
    return null;
  }
  const isRelativeExpression = isExpression || parsedDate.type === "expression";
  const resolvedDate = precision === "day" ? parsedDate.absoluteDate : _getPreciseIsoString(
    parsedDate.absoluteDate,
    timeOffset,
    position,
    isRelativeExpression
  );
  return isRelativeExpression ? { ...parsedDate, absoluteDate: resolvedDate, type: "expression" } : { ...parsedDate, absoluteDate: resolvedDate, value: resolvedDate };
};
function _getPreciseIsoString(isoString, timeOffset, position, isExpression) {
  if (!timeOffset) {
    return isoString;
  }
  const offsetDate = (0, import_helpers.isSimpleTimeOffset)(timeOffset) ? timeOffset : timeOffset[position ?? "from"];
  const date = new import_tz.TZDateMini(isoString, (0, import_user_preferences.getTimezone)());
  if (isExpression) {
    return date.toISOString();
  }
  switch (offsetDate) {
    case "startOfDay":
      return (0, import_date_fns.startOfDay)(date).toISOString();
    case "endOfDay":
      return (0, import_date_fns.endOfDay)(date).toISOString();
    default: {
      if (isoString === offsetDate) {
        return isoString;
      }
      const parsedOffsetDate = (0, import_units.parseTimeAsTimeValue)(offsetDate);
      if (parsedOffsetDate?.type === "expression") {
        return parsedOffsetDate.absoluteDate;
      }
      const precision = (0, import_date_time.getSecondFractions)(offsetDate);
      return (0, import_date_time.applyInitialTime)(isoString, offsetDate, precision);
    }
  }
}
const convertDateOnlyToDateTime = (isoString, type) => {
  if (type === "date") {
    const newDate = new import_tz.TZDateMini(isoString ?? "", (0, import_user_preferences.getTimezone)());
    if (!isNaN(newDate.getTime())) {
      newDate.setMinutes(newDate.getMinutes() + newDate.getTimezoneOffset());
      return newDate.toISOString();
    }
  }
  return isoString ?? "";
};
function toCalendarValue(value, precision) {
  if (!value || value.absoluteDate === "") {
    return null;
  }
  if (precision === "day") {
    const converted = convertDateOnlyToDateTime(value.absoluteDate, "date");
    if (converted) {
      return { absoluteDate: converted, value: converted, type: "iso8601" };
    }
  }
  return value;
}
function filterOutGrailLimit(value, limit) {
  const isOutsideGrailLimit = (0, import_timeframe_selection.isOutsideRange)(
    value,
    import_date_time.GRAIL_MIN_DATE,
    import_date_time.GRAIL_MAX_DATE
  );
  return (limit === import_date_time.GRAIL_MIN_DATE || limit === import_date_time.GRAIL_MAX_DATE) && !isOutsideGrailLimit ? void 0 : limit;
}
