// @dynatrace/strato-components/esm/charts/histogram/components/overlays/HistogramExplore/HistogramTooltip.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { useIntl } from "react-intl";
import { wrapper } from "./HistogramTooltip.sty.js";
import { HistogramTooltipWrapper } from "./HistogramTooltipWrapper.js";
import { Divider } from "../../../../../layouts/divider/Divider.js";
import { ChartActionsMenu } from "../../../../core/components/series-action/ChartActionsMenu.js";
import { useChartActionsMenu } from "../../../../core/components/series-action/hooks/useChartActionsMenu.js";
import { getDefaultAction } from "../../../../core/components/series-action/hooks/useDefaultAction.js";
import { InteractiveChartTooltip } from "../../../../core/components/tooltip/InteractiveChartTooltip/components/InteractiveChartTooltip.js";
import { SeriesBasedContent } from "../../../../core/components/tooltip/InteractiveChartTooltip/contents/SeriesBasedContent.js";
import { CategoricalTooltipHeader } from "../../../../core/components/tooltip/InteractiveChartTooltip/headers/CategoricalTooltipHeader.js";
import { useTooltipSeriesActions } from "../../../../core/components/tooltip/shared/hooks/useTooltipSeriesActions.js";
import { useHistogramFilteredData } from "../../../hooks/useHistogramFilteredData.js";
import { useHistogramTooltip } from "../../../hooks/useHistogramTooltip.js";
import { useHistogramTruncationMode } from "../../../hooks/useHistogramTruncationMode.js";
import { useHistogramXFormatter } from "../../../hooks/useHistogramXFormatter.js";
import { useHistogramYFormatter } from "../../../hooks/useHistogramYFormatter.js";
import { useSeriesActionClickHandler } from "../../../hooks/useSeriesActionClickHandler.js";
import { useXAxisConfig } from "../../../hooks/useXAxisConfig.js";
import { safeGet } from "../../../providers/HistogramDataProvider.js";
import { PREPOSITION } from "../constants.js";
import {
  toTooltipItems
} from "../utils/bins-to-tooltip.js";
import { getHeaderLabel } from "../utils/get-header-label-tooltip.js";
function getHeaderResolution(binsMap, bin, formatter) {
  const { from, to } = safeGet(binsMap, bin);
  return formatter(to - from);
}
const HistogramTooltip = ({
  points,
  closestPoint
}) => {
  const { binsMap, series, seriesMap } = useHistogramFilteredData();
  const { unit: xUnit } = useXAxisConfig();
  const { variant } = useHistogramTooltip();
  const truncationMode = useHistogramTruncationMode();
  const bottomFormatter = useHistogramXFormatter();
  const {
    onLeave: handleMenuLeave,
    onScroll: handleMenuScroll,
    selectedItem,
    updateSelectedItem,
    actionsMenuRef,
    isScrolling
  } = useChartActionsMenu();
  const formatters = useHistogramYFormatter();
  const { units } = useHistogramFilteredData();
  const isSingle = variant === "single";
  const { tooltipItems, closestTooltipItem } = toTooltipItems(
    binsMap,
    formatters,
    isSingle,
    units,
    points,
    closestPoint
  );
  const actionClickHandler = useSeriesActionClickHandler(
    tooltipItems.length > 1
  );
  const seriesActions = useTooltipSeriesActions(
    series,
    (series2) => safeGet(seriesMap, series2).id,
    tooltipItems.find((t) => t.id === selectedItem)?.bin
  );
  const intl = useIntl();
  const showHeaderTooltipItem = !isSingle && tooltipItems.length > 1;
  const showHeaderDivider = tooltipItems.length > 1 && variant === "shared";
  const labelFormatter = (value) => bottomFormatter(value, xUnit);
  const headerLabel = closestPoint ? getHeaderLabel(
    binsMap,
    closestPoint.bin,
    labelFormatter,
    intl.formatMessage(PREPOSITION.to)
  ) : "";
  return /* @__PURE__ */ jsxs(
    HistogramTooltipWrapper,
    {
      tooltipItems,
      onMouseLeave: handleMenuLeave,
      onScrollContent: handleMenuScroll,
      children: [
        /* @__PURE__ */ jsx(InteractiveChartTooltip.Header, { children: closestPoint && /* @__PURE__ */ jsxs(Fragment, { children: [
          /* @__PURE__ */ jsx(
            CategoricalTooltipHeader,
            {
              resolution: getHeaderResolution(
                binsMap,
                closestPoint.bin,
                labelFormatter
              ),
              name: headerLabel,
              item: showHeaderTooltipItem ? {
                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                ...closestTooltipItem,
                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                values: closestTooltipItem.asyncValues()
              } : void 0
            }
          ),
          showHeaderDivider && /* @__PURE__ */ jsx(
            "div",
            {
              className: wrapper,
              children: /* @__PURE__ */ jsx(Divider, {})
            },
            `divider-${closestTooltipItem?.id}`
          )
        ] }) }),
        tooltipItems && /* @__PURE__ */ jsx(InteractiveChartTooltip.Body, { children: tooltipItems.map((item) => () => {
          const isOpen = selectedItem === item.id;
          return /* @__PURE__ */ jsx(
            ChartActionsMenu,
            {
              isOpen,
              actions: isOpen ? seriesActions(item) : void 0,
              defaultActions: getDefaultAction(item),
              ref: actionsMenuRef,
              onActionClick: actionClickHandler,
              children: /* @__PURE__ */ jsx(
                SeriesBasedContent,
                {
                  item: { ...item, values: item.asyncValues() },
                  onMouseHover: () => updateSelectedItem(item.id),
                  truncationMode,
                  isScrolling
                }
              )
            },
            `${item.id}`
          );
        }) })
      ]
    }
  );
};
export {
  HistogramTooltip
};
