// @dynatrace/strato-components/esm/charts/core/components/legend/range/components/RangeLegendItemLabel/RangeLegendItemLabel.js
import { jsx } from "react/jsx-runtime";
import { isUndefined } from "lodash-es";
import { legendLabel } from "./RangeLegendItemLabel.sty.js";
import { ORIENTATION } from "../../../../../constants.js";
import {
  DEFAULT_LABEL_FONT_SIZE,
  HOVERED_LABEL_FONT_WEIGHT,
  LABEL_FONT_COLOR_HOVERED,
  LABEL_HOVERED_PADDING
} from "../../constants.js";
import { useOrientation } from "../../context/RangeLegendOrientationContext.js";
import { useRangeLegendHoveredMarkerId } from "../../hooks/store/selectors.js";
import { useLegendIntermediateLabelCollisionDetector } from "../../hooks/use-legend-intermediate-label-collision-detector.js";
const RangeLegendItemLabel = (props) => {
  const { alignment, endPosition, labelId, startPosition, text } = props;
  const { orientation } = useOrientation();
  const isVertical = orientation === ORIENTATION.Vertical;
  const isHorizontal = orientation === ORIENTATION.Horizontal;
  const hoveredMarkerId = useRangeLegendHoveredMarkerId();
  const isLabelHovered = labelId === hoveredMarkerId;
  const {
    text: label,
    x,
    y
  } = useLegendIntermediateLabelCollisionDetector(
    text,
    startPosition,
    endPosition,
    alignment,
    !isLabelHovered
  );
  if (isUndefined(x) || isUndefined(y) || !isLabelHovered) {
    return null;
  }
  return /* @__PURE__ */ jsx(
    "text",
    {
      x: x + (isHorizontal ? 0 : LABEL_HOVERED_PADDING),
      y: y + (isVertical ? 0 : LABEL_HOVERED_PADDING),
      fontSize: DEFAULT_LABEL_FONT_SIZE,
      fontWeight: HOVERED_LABEL_FONT_WEIGHT,
      fill: LABEL_FONT_COLOR_HOVERED,
      className: legendLabel,
      role: "listitem",
      "aria-label": `Label ${text}`,
      children: label
    }
  );
};
export {
  RangeLegendItemLabel
};
