// @dynatrace/strato-components/esm/content/chip-group/ChipGroup.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import { forwardRef, useEffect, useReducer, useRef, useState } from "react";
import { chipGroupCSS } from "./ChipGroup.sty.js";
import { ChipGroupContext } from "./contexts/ChipGroupContext.js";
import { InternalChipGroupContext } from "./contexts/InternalChipGroupContext.js";
import { Control } from "./Control.js";
import { HiddenCount } from "./HiddenCount.js";
import { chipGroupReducer } from "./state/chip-group-state-reducer.js";
import { useAriaLabelingProps } from "../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useContainerSize } from "../../core/hooks/useContainerSize.js";
import { useControllableState } from "../../core/hooks/useControllableState.js";
import { useId } from "../../core/hooks/useId.js";
import { useMergeRefs } from "../../core/hooks/useMergeRefs.js";
import { useSafeForwardProps } from "../../core/hooks/useSafeForwardProps.js";
import { Flex } from "../../layouts/flex/Flex.js";
const COMPONENT_NAME = "ChipGroup";
const RootChipGroup = forwardRef(
  (props, forwardedRef) => {
    const {
      children,
      expanded: expandedProp,
      defaultExpanded,
      disabled,
      size,
      onExpandedChange,
      maxVisibleChips = "auto",
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
    const ariaLabelingProps = useAriaLabelingProps(restProps);
    const forwardProps = useSafeForwardProps(restProps, ariaLabelingProps);
    const uniqueId = useId();
    const [state, dispatch] = useReducer(chipGroupReducer, {
      chips: {}
    });
    const containerRef = useRef(null);
    const rootRef = useMergeRefs([
      containerRef,
      forwardedRef
    ]);
    const { width: containerWidth } = useContainerSize(containerRef);
    const [expanded, setExpanded] = useControllableState({
      value: expandedProp,
      defaultValue: defaultExpanded,
      onChange: onExpandedChange
    });
    const [allChipsFit, setAllChipsFit] = useState(false);
    const [hiddenChips, setHiddenChips] = useState([]);
    const [visibleChips, setVisibleChips] = useState([]);
    useEffect(() => {
      const chips = Object.entries(state.chips);
      const chipKeys = chips.map(([uniqueChipKey]) => uniqueChipKey);
      if (chips.length === 0) {
        setVisibleChips([]);
        return;
      }
      if (chips.some(([, chip]) => !chip.width)) {
        return;
      }
      if (expanded) {
        setHiddenChips([]);
        setVisibleChips(chipKeys);
        if (!containerRef.current?.firstChild) {
          setAllChipsFit(true);
          return;
        }
        let wrap = false;
        let child = containerRef.current.firstChild;
        const rafId = requestAnimationFrame(() => {
          do {
            if (child.classList.contains("chip") && child.offsetTop > 0) {
              wrap = true;
            }
            child = child.nextSibling;
          } while (!wrap && child);
          setAllChipsFit(!wrap);
          setTimeout(() => {
            const controlButton = containerRef.current?.querySelector(
              `button[aria-controls="${uniqueId}"]`
            );
            if (controlButton) {
              controlButton.focus();
            }
          }, 0);
        });
        return () => {
          cancelAnimationFrame(rafId);
        };
      }
      const spacing = size === "condensed" ? 6 : 8;
      let totalChipsWidth = 0;
      let setHidden = true;
      let _hiddenChipsIndex = chips.length;
      for (const [index, [, value]] of Array.from(chips).entries()) {
        totalChipsWidth += value.width || 0;
        const chipFits = index === 0 || Math.floor(totalChipsWidth + index * spacing) <= Math.round(containerWidth);
        const chipAndControlFits = index === 0 || totalChipsWidth + (index + 1) * spacing + (state.control?.width || 0) <= containerWidth;
        const visible = index === chips.length - 1 && (state.control?.show ?? "auto") === "auto" ? chipFits : chipAndControlFits;
        if (!visible && setHidden) {
          setHidden = false;
          _hiddenChipsIndex = index;
        }
      }
      const effectiveHiddenChipsIndex = maxVisibleChips !== "auto" && maxVisibleChips >= 0 ? Math.min(_hiddenChipsIndex, maxVisibleChips) : _hiddenChipsIndex;
      const allFit = effectiveHiddenChipsIndex === chips.length;
      setAllChipsFit(allFit);
      setHiddenChips(chipKeys.slice(effectiveHiddenChipsIndex));
      setVisibleChips(chipKeys.slice(0, effectiveHiddenChipsIndex));
    }, [containerWidth, expanded, size, state, maxVisibleChips, uniqueId]);
    return /* @__PURE__ */ jsx(
      ChipGroupContext.Provider,
      {
        value: {
          visibleKeys: visibleChips,
          hiddenKeys: hiddenChips
        },
        children: /* @__PURE__ */ jsx(
          InternalChipGroupContext.Provider,
          {
            value: {
              chipGroupId: uniqueId,
              disabled,
              size,
              expanded,
              allChipsFit,
              dispatch,
              setExpanded
            },
            children: /* @__PURE__ */ jsx(
              Flex,
              {
                ref: rootRef,
                id: uniqueId,
                className: clsx(
                  chipGroupCSS({ size, expanded }),
                  consumerClassName
                ),
                style: consumerStyle,
                "data-testid": dataTestId,
                "data-dtrum-mask": dataDtrumMask,
                "data-dtrum-allow": dataDtrumAllow,
                ...ariaLabelingProps,
                ...behavioralTrackingProps,
                ...forwardProps,
                "data-dt-component": COMPONENT_NAME,
                children
              }
            )
          }
        )
      }
    );
  }
);
RootChipGroup.displayName = COMPONENT_NAME;
const ChipGroup = Object.assign(RootChipGroup, {
  Control,
  HiddenCount
});
export {
  ChipGroup
};
