// @dynatrace/strato-components/esm/content/chip-group/Control.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import {
  useEffect,
  useMemo,
  useRef
} from "react";
import { useIntl } from "react-intl";
import * as styles from "./Control.sty.js";
import { useChipGroupContext } from "./hooks/useChipGroupContext.js";
import { useInternalChipGroupContext } from "./hooks/useInternalChipGroupContext.js";
import { ChipGroupActions } from "./state/chip-group-state-reducer.js";
import { useContainerSize } from "../../core/hooks/useContainerSize.js";
import { useFocusRing } from "../../core/styles/useFocusRing.js";
import { ProgressCircle } from "../progress/ProgressCircle.js";
const Control = (props) => {
  const {
    display,
    count,
    loading = false,
    onClick,
    children,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId
  } = props;
  const intl = useIntl();
  const { focusProps, focusClassName, isFocusVisible } = useFocusRing({
    color: "neutral"
  });
  const { hiddenKeys, visibleKeys } = useChipGroupContext();
  const internalChipGroupContext = useInternalChipGroupContext();
  const { chipGroupId, expanded, allChipsFit, setExpanded, dispatch } = internalChipGroupContext || {};
  const controlRef = useRef(null);
  const { width: controlWidth } = useContainerSize(controlRef);
  const width = Math.ceil(controlWidth);
  const label = useMemo(() => {
    const moreOrLess = expanded && display !== "more" ? "less" : "more";
    const hiddenChipsCount = count === true && hiddenKeys.length > 0 ? hiddenKeys.length : count;
    const showCount = typeof hiddenChipsCount === "number" && moreOrLess === "more";
    return intl.formatMessage(
      {
        defaultMessage: "Show {count} {moreOrLess}",
        id: "ARdI7d8cJxHnUayP",
        description: "Label for the expand / collapse button."
      },
      {
        moreOrLess: expanded && display !== "more" ? "less" : "more",
        count: showCount ? hiddenChipsCount : void 0
      }
    );
  }, [count, display, expanded, hiddenKeys, intl]);
  useEffect(() => {
    dispatch?.({
      type: ChipGroupActions.ADDCONTROL,
      payload: { width }
    });
    return () => {
      dispatch?.({
        type: ChipGroupActions.REMOVECONTROL,
        payload: {}
      });
    };
  }, []);
  useEffect(() => {
    dispatch?.({
      type: ChipGroupActions.UPDATECONTROL,
      payload: {
        width: width === 0 ? void 0 : width,
        show: display !== void 0 ? "always" : "auto"
      }
    });
  }, [dispatch, display, width]);
  const shouldHide = display === void 0 && (allChipsFit || visibleKeys.length === 0) && !loading;
  if (shouldHide) {
    return null;
  }
  return /* @__PURE__ */ jsxs(
    "button",
    {
      "data-testid": dataTestId,
      ref: controlRef,
      onClick: (event) => {
        onClick?.(event);
        setExpanded?.((prev) => display === "more" ? true : !prev);
      },
      "aria-controls": chipGroupId,
      "aria-expanded": expanded,
      className: clsx(
        styles.expandButton({ isFocusVisible }),
        focusClassName,
        consumerClassName
      ),
      style: consumerStyle,
      ...focusProps,
      children: [
        loading && /* @__PURE__ */ jsx(
          ProgressCircle,
          {
            size: "small",
            "aria-label": intl.formatMessage({
              defaultMessage: "Loading",
              id: "cycz+bYG+VH7Of4e",
              description: "Label for the loading state."
            })
          }
        ),
        children ?? label
      ]
    }
  );
};
export {
  Control
};
