// @dynatrace-sdk/client-app-settings-v2/cjs/index.js
/**
 * @license
 * Copyright 2023-2025 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/client/app-settings-v2/src/index.ts
var src_exports = {};
__export(src_exports, {
  ApiClientError: () => import_shared_errors3.ApiClientError,
  ApiGatewayError: () => import_shared_errors3.ApiGatewayError,
  AppSettingsObjectsClient: () => AppSettingsObjectsClient,
  AppSettingsPermissionsListItem: () => AppSettingsPermissionsListItem,
  ClientRequestError: () => import_shared_errors3.ClientRequestError,
  ConflictingResource: () => ConflictingResource,
  EffectivePermissionGranted: () => EffectivePermissionGranted,
  ErrorResponseError: () => ErrorResponseError,
  GeneralError: () => GeneralError,
  GetAppSettingsPermissionByObjectIdAndAccessorIdPathAccessorType: () => GetAppSettingsPermissionByObjectIdAndAccessorIdPathAccessorType,
  IdentityType: () => IdentityType,
  InvalidResponseError: () => import_shared_errors3.InvalidResponseError,
  ResourceContextOperationsItem: () => ResourceContextOperationsItem,
  ServiceUnavailable: () => ServiceUnavailable,
  SinglePermissionRequestPermission: () => SinglePermissionRequestPermission,
  TooManyRequests: () => TooManyRequests,
  appSettingsObjectsClient: () => appSettingsObjectsClient,
  isApiClientError: () => import_shared_errors3.isApiClientError,
  isApiGatewayError: () => import_shared_errors3.isApiGatewayError,
  isClientRequestError: () => import_shared_errors3.isClientRequestError,
  isConflictingResource: () => isConflictingResource,
  isErrorResponseError: () => isErrorResponseError,
  isGeneralError: () => isGeneralError,
  isInvalidResponseError: () => import_shared_errors3.isInvalidResponseError,
  isServiceUnavailable: () => isServiceUnavailable,
  isTooManyRequests: () => isTooManyRequests
});
module.exports = __toCommonJS(src_exports);

// packages/client/app-settings-v2/src/lib/apis/app-settings-objects-api.ts
var import_http_client = require("@dynatrace-sdk/http-client");
var import_shared_client_utils = require("@dynatrace-sdk/shared-client-utils");
var import_shared_errors2 = require("@dynatrace-sdk/shared-errors");

// packages/client/app-settings-v2/src/lib/error-envelopes/error-response-error.ts
var import_shared_errors = require("@dynatrace-sdk/shared-errors");
var ErrorResponseError = class extends import_shared_errors.ClientRequestError {
  isErrorResponseError = true;
};
function isErrorResponseError(e) {
  return e?.isApiClientError === true && e?.isClientRequestError === true && e?.isErrorResponseError === true && e instanceof Error;
}

// packages/client/app-settings-v2/src/lib/error-envelopes/responses/conflicting-resource.ts
var ConflictingResource = class extends ErrorResponseError {
  isConflictingResource = true;
  name = "ConflictingResource";
};
function isConflictingResource(e) {
  return e?.isApiClientError === true && e?.isClientRequestError === true && e?.isErrorResponseError === true && e?.isConflictingResource === true && e instanceof Error;
}

// packages/client/app-settings-v2/src/lib/error-envelopes/responses/general-error.ts
var GeneralError = class extends ErrorResponseError {
  isGeneralError = true;
  name = "GeneralError";
};
function isGeneralError(e) {
  return e?.isApiClientError === true && e?.isClientRequestError === true && e?.isErrorResponseError === true && e?.isGeneralError === true && e instanceof Error;
}

// packages/client/app-settings-v2/src/lib/error-envelopes/responses/service-unavailable.ts
var ServiceUnavailable = class extends ErrorResponseError {
  isServiceUnavailable = true;
  name = "ServiceUnavailable";
};
function isServiceUnavailable(e) {
  return e?.isApiClientError === true && e?.isClientRequestError === true && e?.isErrorResponseError === true && e?.isServiceUnavailable === true && e instanceof Error;
}

// packages/client/app-settings-v2/src/lib/error-envelopes/responses/too-many-requests.ts
var TooManyRequests = class extends ErrorResponseError {
  isTooManyRequests = true;
  name = "TooManyRequests";
};
function isTooManyRequests(e) {
  return e?.isApiClientError === true && e?.isClientRequestError === true && e?.isErrorResponseError === true && e?.isTooManyRequests === true && e instanceof Error;
}

// packages/client/app-settings-v2/src/lib/apis/app-settings-objects-api.ts
var AppSettingsObjectsClient = class {
  httpClient;
  shouldTransformDates = true;
  /**
   * @param {HttpClient} httpClientImplementation - You can provide custom http client as a parameter to constructor of a AppSettingsObjectsClient. Custom http client should implement {HttpClient} interface.
   * @param {ApiClientOptions} [options] - Optional configuration for SDK Client.
   * @example
   * const appSettingsObjectsClientNodeJs = new AppSettingsObjectsClient(yourCustomImplementation);
   */
  constructor(httpClientImplementation, options) {
    this.httpClient = httpClientImplementation;
    this.shouldTransformDates = options?.transformDates ?? true;
  }
  /**
   * Gets the specified settings object
   *
   * **Required scope:** app-settings:objects:read<br/>**Required permission:** app-settings:objects:read<br/><br/>Gets the specified settings object. Properties of type **secret** will be included in plain text if the call originates from a serverless function of your app; they will have irreversibly masked values otherwise. This protects these secrets from leaking to users of your app or other third parties.
   *
   * @returns {Promise<AppSettingsObject>} Success
   *
   * @example <caption>Code example</caption>
   * import { appSettingsObjectsClient } from "@dynatrace-sdk/client-app-settings-v2";
   *
   * const data =
   *   await appSettingsObjectsClient.getAppSettingsObjectByObjectId(
   *     { objectId: "..." },
   *   );
   *
   * @throws {ErrorResponseError} Failed. The input is invalid. | Failed. Forbidden. | No object available for the given objectId
   * @throws {TooManyRequests} Failed. Too many requests.
   * @throws {ServiceUnavailable} Failed. Service unavailable.
   * @throws {GeneralError} Error.
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async getAppSettingsObjectByObjectId(config) {
    if (!config) {
      throw new import_shared_errors2.ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const query = (0, import_shared_client_utils.toQueryString)({ "admin-access": config.adminAccess });
    try {
      const response = await this.httpClient.send({
        url: `/platform/app-settings/v2/objects/${config.objectId}${query}`,
        method: "GET",
        headers: {
          Accept: "application/json; charset=utf-8, application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, [
          ["modificationInfo.createdTime", { datetime: true }],
          ["modificationInfo.lastModifiedTime", { datetime: true }]
        ]) : responseValue;
      } catch (err) {
        throw new import_shared_errors2.InvalidResponseError(
          `AppSettingsObjectsClient.getAppSettingsObjectByObjectId:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if ((0, import_http_client.isHttpClientAbortError)(e) || (0, import_shared_errors2.isInvalidResponseError)(e)) {
        throw e;
      }
      if (!(0, import_http_client.isHttpClientResponseError)(e)) {
        throw new import_shared_errors2.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await (0, import_shared_client_utils.apiGatewayErrorsHandler)(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ErrorResponseError(
              `400`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. The input is invalid."),
              e
            );
          } catch (err) {
            if (err instanceof ErrorResponseError) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.getAppSettingsObjectByObjectId:400`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ErrorResponseError(
              `403`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. Forbidden."),
              e
            );
          } catch (err) {
            if (err instanceof ErrorResponseError) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.getAppSettingsObjectByObjectId:403`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ErrorResponseError(
              `404`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "No object available for the given objectId"),
              e
            );
          } catch (err) {
            if (err instanceof ErrorResponseError) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.getAppSettingsObjectByObjectId:404`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 429: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new TooManyRequests(
              `429`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. Too many requests."),
              e
            );
          } catch (err) {
            if (err instanceof TooManyRequests) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.getAppSettingsObjectByObjectId:429`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 503: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ServiceUnavailable(
              `503`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. Service unavailable."),
              e
            );
          } catch (err) {
            if (err instanceof ServiceUnavailable) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.getAppSettingsObjectByObjectId:503`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        default: {
          if (response.status >= 400 && response.status <= 499) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
              throw new GeneralError(
                `${response.status}`,
                response,
                errorBody,
                (0, import_shared_client_utils.getErrorMessage)(errorBody, "Error."),
                e
              );
            } catch (err) {
              if (err instanceof GeneralError) {
                throw err;
              }
              throw new import_shared_errors2.InvalidResponseError(
                `AppSettingsObjectsClient.getAppSettingsObjectByObjectId:${response.status}`,
                err,
                responseValue,
                "ErrorResponse",
                void 0,
                response
              );
            }
          } else if (response.status >= 500 && response.status <= 599) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
              throw new GeneralError(
                `${response.status}`,
                response,
                errorBody,
                (0, import_shared_client_utils.getErrorMessage)(errorBody, "Error."),
                e
              );
            } catch (err) {
              if (err instanceof GeneralError) {
                throw err;
              }
              throw new import_shared_errors2.InvalidResponseError(
                `AppSettingsObjectsClient.getAppSettingsObjectByObjectId:${response.status}`,
                err,
                responseValue,
                "ErrorResponse",
                void 0,
                response
              );
            }
          } else {
            try {
              const responseValue = await response.body("json");
              throw new import_shared_errors2.ClientRequestError(
                `${response.status}`,
                response,
                responseValue,
                (0, import_shared_client_utils.getErrorMessage)(
                  responseValue,
                  `Unexpected api response: code=${response.status} body="${responseValue}"`
                )
              );
            } catch (error) {
              if (error instanceof import_shared_errors2.ClientRequestError) {
                throw error;
              }
              const responseValue = await response.body("text").catch(() => "");
              throw new import_shared_errors2.InvalidResponseError(
                `${response.status}`,
                error,
                responseValue,
                "json",
                "Unexpected api response",
                response
              );
            }
          }
        }
      }
    }
  }
  /**
   * Updates an existing settings object
   *
   * **Required scope:** app-settings:objects:write<br/>**Required permission:** app-settings:objects:write<br/><br/>Updates an existing settings object with new values. To update a property of the `secret` type you need to pass the new value unmasked. To keep the current value, send the current masked secret. You can obtain it via [GET an object](#getappsettingsobjects) endpoint.
   *
   * Some schemas don't allow passing of the masked secret. In that case you need to send the unmasked secret with every update of the object.
   *
   * @returns {Promise<AppSettingsUpdateResponse>} Success
   *
   * @example <caption>Code example</caption>
   * import { appSettingsObjectsClient } from "@dynatrace-sdk/client-app-settings-v2";
   *
   * const data =
   *   await appSettingsObjectsClient.putAppSettingsObjectByObjectId(
   *     {
   *       objectId: "...",
   *       optimisticLockingVersion: "...",
   *       body: { value: {} },
   *     },
   *   );
   *
   * @throws {ErrorResponseError} Failed. The input is invalid. | Failed. Forbidden. | Failed. The requested resource doesn't exist.
   * @throws {ConflictingResource} Failed. Conflicting resource.
   * @throws {TooManyRequests} Failed. Too many requests.
   * @throws {ServiceUnavailable} Failed. Service unavailable.
   * @throws {GeneralError} Error.
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async putAppSettingsObjectByObjectId(config) {
    if (!config) {
      throw new import_shared_errors2.ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const encodedBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformRequest)(config.body, []) : config.body;
    const query = (0, import_shared_client_utils.toQueryString)({
      "optimistic-locking-version": config.optimisticLockingVersion,
      "admin-access": config.adminAccess
    });
    try {
      const response = await this.httpClient.send({
        url: `/platform/app-settings/v2/objects/${config.objectId}${query}`,
        method: "PUT",
        requestBodyType: "json",
        body: encodedBody,
        headers: {
          "Content-Type": "application/json; charset=utf-8",
          Accept: "application/json; charset=utf-8, application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
      } catch (err) {
        throw new import_shared_errors2.InvalidResponseError(
          `AppSettingsObjectsClient.putAppSettingsObjectByObjectId:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if ((0, import_http_client.isHttpClientAbortError)(e) || (0, import_shared_errors2.isInvalidResponseError)(e)) {
        throw e;
      }
      if (!(0, import_http_client.isHttpClientResponseError)(e)) {
        throw new import_shared_errors2.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await (0, import_shared_client_utils.apiGatewayErrorsHandler)(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ErrorResponseError(
              `400`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. The input is invalid."),
              e
            );
          } catch (err) {
            if (err instanceof ErrorResponseError) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.putAppSettingsObjectByObjectId:400`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ErrorResponseError(
              `403`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. Forbidden."),
              e
            );
          } catch (err) {
            if (err instanceof ErrorResponseError) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.putAppSettingsObjectByObjectId:403`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ErrorResponseError(
              `404`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. The requested resource doesn't exist."),
              e
            );
          } catch (err) {
            if (err instanceof ErrorResponseError) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.putAppSettingsObjectByObjectId:404`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 409: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ConflictingResource(
              `409`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. Conflicting resource."),
              e
            );
          } catch (err) {
            if (err instanceof ConflictingResource) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.putAppSettingsObjectByObjectId:409`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 429: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new TooManyRequests(
              `429`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. Too many requests."),
              e
            );
          } catch (err) {
            if (err instanceof TooManyRequests) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.putAppSettingsObjectByObjectId:429`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 503: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ServiceUnavailable(
              `503`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. Service unavailable."),
              e
            );
          } catch (err) {
            if (err instanceof ServiceUnavailable) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.putAppSettingsObjectByObjectId:503`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        default: {
          if (response.status >= 400 && response.status <= 499) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
              throw new GeneralError(
                `${response.status}`,
                response,
                errorBody,
                (0, import_shared_client_utils.getErrorMessage)(errorBody, "Error."),
                e
              );
            } catch (err) {
              if (err instanceof GeneralError) {
                throw err;
              }
              throw new import_shared_errors2.InvalidResponseError(
                `AppSettingsObjectsClient.putAppSettingsObjectByObjectId:${response.status}`,
                err,
                responseValue,
                "ErrorResponse",
                void 0,
                response
              );
            }
          } else if (response.status >= 500 && response.status <= 599) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
              throw new GeneralError(
                `${response.status}`,
                response,
                errorBody,
                (0, import_shared_client_utils.getErrorMessage)(errorBody, "Error."),
                e
              );
            } catch (err) {
              if (err instanceof GeneralError) {
                throw err;
              }
              throw new import_shared_errors2.InvalidResponseError(
                `AppSettingsObjectsClient.putAppSettingsObjectByObjectId:${response.status}`,
                err,
                responseValue,
                "ErrorResponse",
                void 0,
                response
              );
            }
          } else {
            try {
              const responseValue = await response.body("json");
              throw new import_shared_errors2.ClientRequestError(
                `${response.status}`,
                response,
                responseValue,
                (0, import_shared_client_utils.getErrorMessage)(
                  responseValue,
                  `Unexpected api response: code=${response.status} body="${responseValue}"`
                )
              );
            } catch (error) {
              if (error instanceof import_shared_errors2.ClientRequestError) {
                throw error;
              }
              const responseValue = await response.body("text").catch(() => "");
              throw new import_shared_errors2.InvalidResponseError(
                `${response.status}`,
                error,
                responseValue,
                "json",
                "Unexpected api response",
                response
              );
            }
          }
        }
      }
    }
  }
  /**
   * Deletes the specified settings object
   *
   * **Required scope:** app-settings:objects:write<br/>**Required permission:** app-settings:objects:write
   *
   * @example <caption>Code example</caption>
   * import { appSettingsObjectsClient } from "@dynatrace-sdk/client-app-settings-v2";
   *
   * const data =
   *   await appSettingsObjectsClient.deleteAppSettingsObjectByObjectId(
   *     { objectId: "...", optimisticLockingVersion: "..." },
   *   );
   *
   * @throws {ErrorResponseError} Failed. The input is invalid. | Failed. Forbidden. | Failed. The requested resource doesn't exist.
   * @throws {ConflictingResource} Failed. Conflicting resource.
   * @throws {TooManyRequests} Failed. Too many requests.
   * @throws {ServiceUnavailable} Failed. Service unavailable.
   * @throws {GeneralError} Error.
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async deleteAppSettingsObjectByObjectId(config) {
    if (!config) {
      throw new import_shared_errors2.ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const query = (0, import_shared_client_utils.toQueryString)({
      "optimistic-locking-version": config.optimisticLockingVersion,
      "admin-access": config.adminAccess
    });
    try {
      await this.httpClient.send({
        url: `/platform/app-settings/v2/objects/${config.objectId}${query}`,
        method: "DELETE",
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [204].includes(status);
        }
      });
      return;
    } catch (e) {
      if ((0, import_http_client.isHttpClientAbortError)(e)) {
        throw e;
      }
      if (!(0, import_http_client.isHttpClientResponseError)(e)) {
        throw new import_shared_errors2.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await (0, import_shared_client_utils.apiGatewayErrorsHandler)(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ErrorResponseError(
              `400`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. The input is invalid."),
              e
            );
          } catch (err) {
            if (err instanceof ErrorResponseError) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.deleteAppSettingsObjectByObjectId:400`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ErrorResponseError(
              `403`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. Forbidden."),
              e
            );
          } catch (err) {
            if (err instanceof ErrorResponseError) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.deleteAppSettingsObjectByObjectId:403`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ErrorResponseError(
              `404`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. The requested resource doesn't exist."),
              e
            );
          } catch (err) {
            if (err instanceof ErrorResponseError) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.deleteAppSettingsObjectByObjectId:404`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 409: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ConflictingResource(
              `409`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. Conflicting resource."),
              e
            );
          } catch (err) {
            if (err instanceof ConflictingResource) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.deleteAppSettingsObjectByObjectId:409`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 429: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new TooManyRequests(
              `429`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. Too many requests."),
              e
            );
          } catch (err) {
            if (err instanceof TooManyRequests) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.deleteAppSettingsObjectByObjectId:429`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 503: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ServiceUnavailable(
              `503`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. Service unavailable."),
              e
            );
          } catch (err) {
            if (err instanceof ServiceUnavailable) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.deleteAppSettingsObjectByObjectId:503`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        default: {
          if (response.status >= 400 && response.status <= 499) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
              throw new GeneralError(
                `${response.status}`,
                response,
                errorBody,
                (0, import_shared_client_utils.getErrorMessage)(errorBody, "Error."),
                e
              );
            } catch (err) {
              if (err instanceof GeneralError) {
                throw err;
              }
              throw new import_shared_errors2.InvalidResponseError(
                `AppSettingsObjectsClient.deleteAppSettingsObjectByObjectId:${response.status}`,
                err,
                responseValue,
                "ErrorResponse",
                void 0,
                response
              );
            }
          } else if (response.status >= 500 && response.status <= 599) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
              throw new GeneralError(
                `${response.status}`,
                response,
                errorBody,
                (0, import_shared_client_utils.getErrorMessage)(errorBody, "Error."),
                e
              );
            } catch (err) {
              if (err instanceof GeneralError) {
                throw err;
              }
              throw new import_shared_errors2.InvalidResponseError(
                `AppSettingsObjectsClient.deleteAppSettingsObjectByObjectId:${response.status}`,
                err,
                responseValue,
                "ErrorResponse",
                void 0,
                response
              );
            }
          } else {
            try {
              const responseValue = await response.body("json");
              throw new import_shared_errors2.ClientRequestError(
                `${response.status}`,
                response,
                responseValue,
                (0, import_shared_client_utils.getErrorMessage)(
                  responseValue,
                  `Unexpected api response: code=${response.status} body="${responseValue}"`
                )
              );
            } catch (error) {
              if (error instanceof import_shared_errors2.ClientRequestError) {
                throw error;
              }
              const responseValue = await response.body("text").catch(() => "");
              throw new import_shared_errors2.InvalidResponseError(
                `${response.status}`,
                error,
                responseValue,
                "json",
                "Unexpected api response",
                response
              );
            }
          }
        }
      }
    }
  }
  /**
   * **Required scope:** app-settings:objects:write<br/>**Required permission:** app-settings:objects:write<br/><br/>Transfer ownership of the object, only the owner or the main admin is allowed to transfer the ownership (IAM permission needed).
   *
   * @example <caption>Code example</caption>
   * import { appSettingsObjectsClient } from "@dynatrace-sdk/client-app-settings-v2";
   *
   * const data =
   *   await appSettingsObjectsClient.postAppSettingsOwnershipByObjectId(
   *     {
   *       objectId: "...",
   *       body: { newOwner: { type: "user" } },
   *     },
   *   );
   *
   * @throws {ErrorResponseError} No object available for the given objectId
   * @throws {TooManyRequests} Failed. Too many requests.
   * @throws {ServiceUnavailable} Failed. Service unavailable.
   * @throws {GeneralError} Error.
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async postAppSettingsOwnershipByObjectId(config) {
    if (!config) {
      throw new import_shared_errors2.ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const encodedBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformRequest)(config.body, []) : config.body;
    const query = (0, import_shared_client_utils.toQueryString)({ "admin-access": config.adminAccess });
    try {
      await this.httpClient.send({
        url: `/platform/app-settings/v2/objects/${config.objectId}:transfer-ownership${query}`,
        method: "POST",
        requestBodyType: "json",
        body: encodedBody,
        headers: {
          "Content-Type": "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [204].includes(status);
        }
      });
      return;
    } catch (e) {
      if ((0, import_http_client.isHttpClientAbortError)(e)) {
        throw e;
      }
      if (!(0, import_http_client.isHttpClientResponseError)(e)) {
        throw new import_shared_errors2.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await (0, import_shared_client_utils.apiGatewayErrorsHandler)(response);
      switch (response.status) {
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ErrorResponseError(
              `404`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "No object available for the given objectId"),
              e
            );
          } catch (err) {
            if (err instanceof ErrorResponseError) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.postAppSettingsOwnershipByObjectId:404`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 429: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new TooManyRequests(
              `429`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. Too many requests."),
              e
            );
          } catch (err) {
            if (err instanceof TooManyRequests) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.postAppSettingsOwnershipByObjectId:429`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 503: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ServiceUnavailable(
              `503`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. Service unavailable."),
              e
            );
          } catch (err) {
            if (err instanceof ServiceUnavailable) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.postAppSettingsOwnershipByObjectId:503`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        default: {
          if (response.status >= 400 && response.status <= 499) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
              throw new GeneralError(
                `${response.status}`,
                response,
                errorBody,
                (0, import_shared_client_utils.getErrorMessage)(errorBody, "Error."),
                e
              );
            } catch (err) {
              if (err instanceof GeneralError) {
                throw err;
              }
              throw new import_shared_errors2.InvalidResponseError(
                `AppSettingsObjectsClient.postAppSettingsOwnershipByObjectId:${response.status}`,
                err,
                responseValue,
                "ErrorResponse",
                void 0,
                response
              );
            }
          } else if (response.status >= 500 && response.status <= 599) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
              throw new GeneralError(
                `${response.status}`,
                response,
                errorBody,
                (0, import_shared_client_utils.getErrorMessage)(errorBody, "Error."),
                e
              );
            } catch (err) {
              if (err instanceof GeneralError) {
                throw err;
              }
              throw new import_shared_errors2.InvalidResponseError(
                `AppSettingsObjectsClient.postAppSettingsOwnershipByObjectId:${response.status}`,
                err,
                responseValue,
                "ErrorResponse",
                void 0,
                response
              );
            }
          } else {
            try {
              const responseValue = await response.body("json");
              throw new import_shared_errors2.ClientRequestError(
                `${response.status}`,
                response,
                responseValue,
                (0, import_shared_client_utils.getErrorMessage)(
                  responseValue,
                  `Unexpected api response: code=${response.status} body="${responseValue}"`
                )
              );
            } catch (error) {
              if (error instanceof import_shared_errors2.ClientRequestError) {
                throw error;
              }
              const responseValue = await response.body("text").catch(() => "");
              throw new import_shared_errors2.InvalidResponseError(
                `${response.status}`,
                error,
                responseValue,
                "json",
                "Unexpected api response",
                response
              );
            }
          }
        }
      }
    }
  }
  /**
   * **Required scope:** app-settings:objects:read<br/>**Required permission:** app-settings:objects:read<br/><br/>Get current permissions on this object
   *
   * @returns {Promise<AppSettingsAccessorPermissionsList>} Success
   *
   * @example <caption>Code example</caption>
   * import { appSettingsObjectsClient } from "@dynatrace-sdk/client-app-settings-v2";
   *
   * const data =
   *   await appSettingsObjectsClient.getAppSettingsPermissionsByObjectId(
   *     { objectId: "..." },
   *   );
   *
   * @throws {ErrorResponseError} No object available for the given objectId
   * @throws {TooManyRequests} Failed. Too many requests.
   * @throws {ServiceUnavailable} Failed. Service unavailable.
   * @throws {GeneralError} Error.
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async getAppSettingsPermissionsByObjectId(config) {
    if (!config) {
      throw new import_shared_errors2.ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const query = (0, import_shared_client_utils.toQueryString)({ "admin-access": config.adminAccess });
    try {
      const response = await this.httpClient.send({
        url: `/platform/app-settings/v2/objects/${config.objectId}/permissions${query}`,
        method: "GET",
        headers: {
          Accept: "application/json; charset=utf-8, application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
      } catch (err) {
        throw new import_shared_errors2.InvalidResponseError(
          `AppSettingsObjectsClient.getAppSettingsPermissionsByObjectId:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if ((0, import_http_client.isHttpClientAbortError)(e) || (0, import_shared_errors2.isInvalidResponseError)(e)) {
        throw e;
      }
      if (!(0, import_http_client.isHttpClientResponseError)(e)) {
        throw new import_shared_errors2.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await (0, import_shared_client_utils.apiGatewayErrorsHandler)(response);
      switch (response.status) {
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ErrorResponseError(
              `404`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "No object available for the given objectId"),
              e
            );
          } catch (err) {
            if (err instanceof ErrorResponseError) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.getAppSettingsPermissionsByObjectId:404`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 429: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new TooManyRequests(
              `429`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. Too many requests."),
              e
            );
          } catch (err) {
            if (err instanceof TooManyRequests) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.getAppSettingsPermissionsByObjectId:429`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 503: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ServiceUnavailable(
              `503`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. Service unavailable."),
              e
            );
          } catch (err) {
            if (err instanceof ServiceUnavailable) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.getAppSettingsPermissionsByObjectId:503`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        default: {
          if (response.status >= 400 && response.status <= 499) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
              throw new GeneralError(
                `${response.status}`,
                response,
                errorBody,
                (0, import_shared_client_utils.getErrorMessage)(errorBody, "Error."),
                e
              );
            } catch (err) {
              if (err instanceof GeneralError) {
                throw err;
              }
              throw new import_shared_errors2.InvalidResponseError(
                `AppSettingsObjectsClient.getAppSettingsPermissionsByObjectId:${response.status}`,
                err,
                responseValue,
                "ErrorResponse",
                void 0,
                response
              );
            }
          } else if (response.status >= 500 && response.status <= 599) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
              throw new GeneralError(
                `${response.status}`,
                response,
                errorBody,
                (0, import_shared_client_utils.getErrorMessage)(errorBody, "Error."),
                e
              );
            } catch (err) {
              if (err instanceof GeneralError) {
                throw err;
              }
              throw new import_shared_errors2.InvalidResponseError(
                `AppSettingsObjectsClient.getAppSettingsPermissionsByObjectId:${response.status}`,
                err,
                responseValue,
                "ErrorResponse",
                void 0,
                response
              );
            }
          } else {
            try {
              const responseValue = await response.body("json");
              throw new import_shared_errors2.ClientRequestError(
                `${response.status}`,
                response,
                responseValue,
                (0, import_shared_client_utils.getErrorMessage)(
                  responseValue,
                  `Unexpected api response: code=${response.status} body="${responseValue}"`
                )
              );
            } catch (error) {
              if (error instanceof import_shared_errors2.ClientRequestError) {
                throw error;
              }
              const responseValue = await response.body("text").catch(() => "");
              throw new import_shared_errors2.InvalidResponseError(
                `${response.status}`,
                error,
                responseValue,
                "json",
                "Unexpected api response",
                response
              );
            }
          }
        }
      }
    }
  }
  /**
   * **Required scope:** app-settings:objects:write<br/>**Required permission:** app-settings:objects:write<br/><br/>Add the permissions for a single accessor on this object, anyone with read/write permissions is allowed to add more permissions
   *
   * @returns {Promise<AppSettingsAccessorPermissions>} Success
   *
   * @example <caption>Code example</caption>
   * import { appSettingsObjectsClient } from "@dynatrace-sdk/client-app-settings-v2";
   *
   * const data =
   *   await appSettingsObjectsClient.postAppSettingsPermissionByObjectId(
   *     {
   *       objectId: "...",
   *       body: { accessor: { type: "user" }, permissions: {} },
   *     },
   *   );
   *
   * @throws {ErrorResponseError} if accessor id already exists | No object available for the given objectId
   * @throws {TooManyRequests} Failed. Too many requests.
   * @throws {ServiceUnavailable} Failed. Service unavailable.
   * @throws {GeneralError} Error.
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async postAppSettingsPermissionByObjectId(config) {
    if (!config) {
      throw new import_shared_errors2.ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const encodedBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformRequest)(config.body, []) : config.body;
    const query = (0, import_shared_client_utils.toQueryString)({ "admin-access": config.adminAccess });
    try {
      const response = await this.httpClient.send({
        url: `/platform/app-settings/v2/objects/${config.objectId}/permissions${query}`,
        method: "POST",
        requestBodyType: "json",
        body: encodedBody,
        headers: {
          "Content-Type": "application/json",
          Accept: "application/json; charset=utf-8, application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [201].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
      } catch (err) {
        throw new import_shared_errors2.InvalidResponseError(
          `AppSettingsObjectsClient.postAppSettingsPermissionByObjectId:201`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if ((0, import_http_client.isHttpClientAbortError)(e) || (0, import_shared_errors2.isInvalidResponseError)(e)) {
        throw e;
      }
      if (!(0, import_http_client.isHttpClientResponseError)(e)) {
        throw new import_shared_errors2.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await (0, import_shared_client_utils.apiGatewayErrorsHandler)(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ErrorResponseError(
              `400`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "if accessor id already exists"),
              e
            );
          } catch (err) {
            if (err instanceof ErrorResponseError) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.postAppSettingsPermissionByObjectId:400`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ErrorResponseError(
              `404`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "No object available for the given objectId"),
              e
            );
          } catch (err) {
            if (err instanceof ErrorResponseError) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.postAppSettingsPermissionByObjectId:404`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 429: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new TooManyRequests(
              `429`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. Too many requests."),
              e
            );
          } catch (err) {
            if (err instanceof TooManyRequests) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.postAppSettingsPermissionByObjectId:429`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 503: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ServiceUnavailable(
              `503`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. Service unavailable."),
              e
            );
          } catch (err) {
            if (err instanceof ServiceUnavailable) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.postAppSettingsPermissionByObjectId:503`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        default: {
          if (response.status >= 400 && response.status <= 499) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
              throw new GeneralError(
                `${response.status}`,
                response,
                errorBody,
                (0, import_shared_client_utils.getErrorMessage)(errorBody, "Error."),
                e
              );
            } catch (err) {
              if (err instanceof GeneralError) {
                throw err;
              }
              throw new import_shared_errors2.InvalidResponseError(
                `AppSettingsObjectsClient.postAppSettingsPermissionByObjectId:${response.status}`,
                err,
                responseValue,
                "ErrorResponse",
                void 0,
                response
              );
            }
          } else if (response.status >= 500 && response.status <= 599) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
              throw new GeneralError(
                `${response.status}`,
                response,
                errorBody,
                (0, import_shared_client_utils.getErrorMessage)(errorBody, "Error."),
                e
              );
            } catch (err) {
              if (err instanceof GeneralError) {
                throw err;
              }
              throw new import_shared_errors2.InvalidResponseError(
                `AppSettingsObjectsClient.postAppSettingsPermissionByObjectId:${response.status}`,
                err,
                responseValue,
                "ErrorResponse",
                void 0,
                response
              );
            }
          } else {
            try {
              const responseValue = await response.body("json");
              throw new import_shared_errors2.ClientRequestError(
                `${response.status}`,
                response,
                responseValue,
                (0, import_shared_client_utils.getErrorMessage)(
                  responseValue,
                  `Unexpected api response: code=${response.status} body="${responseValue}"`
                )
              );
            } catch (error) {
              if (error instanceof import_shared_errors2.ClientRequestError) {
                throw error;
              }
              const responseValue = await response.body("text").catch(() => "");
              throw new import_shared_errors2.InvalidResponseError(
                `${response.status}`,
                error,
                responseValue,
                "json",
                "Unexpected api response",
                response
              );
            }
          }
        }
      }
    }
  }
  /**
   * **Required scope:** app-settings:objects:read<br/>**Required permission:** app-settings:objects:read<br/><br/>Get current permissions of the accessor on this object
   *
   * @returns {Promise<AppSettingsAccessorPermissions>} Success
   *
   * @example <caption>Code example</caption>
   * import { appSettingsObjectsClient } from "@dynatrace-sdk/client-app-settings-v2";
   *
   * const data =
   *   await appSettingsObjectsClient.getAppSettingsPermissionByObjectIdAndAccessorId(
   *     {
   *       objectId: "...",
   *       accessorType: "user",
   *       accessorId: "...",
   *     },
   *   );
   *
   * @throws {ErrorResponseError} No object available for the given objectId or the accessor doesn't have any permissions on this object
   * @throws {TooManyRequests} Failed. Too many requests.
   * @throws {ServiceUnavailable} Failed. Service unavailable.
   * @throws {GeneralError} Error.
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async getAppSettingsPermissionByObjectIdAndAccessorId(config) {
    if (!config) {
      throw new import_shared_errors2.ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const query = (0, import_shared_client_utils.toQueryString)({ "admin-access": config.adminAccess });
    try {
      const response = await this.httpClient.send({
        url: `/platform/app-settings/v2/objects/${config.objectId}/permissions/${config.accessorType}/${config.accessorId}${query}`,
        method: "GET",
        headers: {
          Accept: "application/json; charset=utf-8, application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
      } catch (err) {
        throw new import_shared_errors2.InvalidResponseError(
          `AppSettingsObjectsClient.getAppSettingsPermissionByObjectIdAndAccessorId:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if ((0, import_http_client.isHttpClientAbortError)(e) || (0, import_shared_errors2.isInvalidResponseError)(e)) {
        throw e;
      }
      if (!(0, import_http_client.isHttpClientResponseError)(e)) {
        throw new import_shared_errors2.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await (0, import_shared_client_utils.apiGatewayErrorsHandler)(response);
      switch (response.status) {
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ErrorResponseError(
              `404`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(
                errorBody,
                "No object available for the given objectId or the accessor doesn't have any permissions on this object"
              ),
              e
            );
          } catch (err) {
            if (err instanceof ErrorResponseError) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.getAppSettingsPermissionByObjectIdAndAccessorId:404`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 429: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new TooManyRequests(
              `429`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. Too many requests."),
              e
            );
          } catch (err) {
            if (err instanceof TooManyRequests) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.getAppSettingsPermissionByObjectIdAndAccessorId:429`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 503: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ServiceUnavailable(
              `503`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. Service unavailable."),
              e
            );
          } catch (err) {
            if (err instanceof ServiceUnavailable) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.getAppSettingsPermissionByObjectIdAndAccessorId:503`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        default: {
          if (response.status >= 400 && response.status <= 499) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
              throw new GeneralError(
                `${response.status}`,
                response,
                errorBody,
                (0, import_shared_client_utils.getErrorMessage)(errorBody, "Error."),
                e
              );
            } catch (err) {
              if (err instanceof GeneralError) {
                throw err;
              }
              throw new import_shared_errors2.InvalidResponseError(
                `AppSettingsObjectsClient.getAppSettingsPermissionByObjectIdAndAccessorId:${response.status}`,
                err,
                responseValue,
                "ErrorResponse",
                void 0,
                response
              );
            }
          } else if (response.status >= 500 && response.status <= 599) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
              throw new GeneralError(
                `${response.status}`,
                response,
                errorBody,
                (0, import_shared_client_utils.getErrorMessage)(errorBody, "Error."),
                e
              );
            } catch (err) {
              if (err instanceof GeneralError) {
                throw err;
              }
              throw new import_shared_errors2.InvalidResponseError(
                `AppSettingsObjectsClient.getAppSettingsPermissionByObjectIdAndAccessorId:${response.status}`,
                err,
                responseValue,
                "ErrorResponse",
                void 0,
                response
              );
            }
          } else {
            try {
              const responseValue = await response.body("json");
              throw new import_shared_errors2.ClientRequestError(
                `${response.status}`,
                response,
                responseValue,
                (0, import_shared_client_utils.getErrorMessage)(
                  responseValue,
                  `Unexpected api response: code=${response.status} body="${responseValue}"`
                )
              );
            } catch (error) {
              if (error instanceof import_shared_errors2.ClientRequestError) {
                throw error;
              }
              const responseValue = await response.body("text").catch(() => "");
              throw new import_shared_errors2.InvalidResponseError(
                `${response.status}`,
                error,
                responseValue,
                "json",
                "Unexpected api response",
                response
              );
            }
          }
        }
      }
    }
  }
  /**
   * **Required scope:** app-settings:objects:write<br/>**Required permission:** app-settings:objects:write<br/><br/>Update the permissions for an existing accessor on this object, anyone with read/write permissions is allowed to update permissions
   *
   * @example <caption>Code example</caption>
   * import { appSettingsObjectsClient } from "@dynatrace-sdk/client-app-settings-v2";
   *
   * const data =
   *   await appSettingsObjectsClient.putAppSettingsPermissionByObjectIdAndAccessorId(
   *     {
   *       objectId: "...",
   *       accessorType: "user",
   *       accessorId: "...",
   *       body: { permissions: {} },
   *     },
   *   );
   *
   * @throws {ErrorResponseError} if permission list is empty, contains unsupported entries or unsupported combinations of entries | No object available for the given objectId or the accessor doesn't have any permissions on this object
   * @throws {TooManyRequests} Failed. Too many requests.
   * @throws {ServiceUnavailable} Failed. Service unavailable.
   * @throws {GeneralError} Error.
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async putAppSettingsPermissionByObjectIdAndAccessorId(config) {
    if (!config) {
      throw new import_shared_errors2.ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const encodedBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformRequest)(config.body, []) : config.body;
    const query = (0, import_shared_client_utils.toQueryString)({ "admin-access": config.adminAccess });
    try {
      await this.httpClient.send({
        url: `/platform/app-settings/v2/objects/${config.objectId}/permissions/${config.accessorType}/${config.accessorId}${query}`,
        method: "PUT",
        requestBodyType: "json",
        body: encodedBody,
        headers: {
          "Content-Type": "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      return;
    } catch (e) {
      if ((0, import_http_client.isHttpClientAbortError)(e)) {
        throw e;
      }
      if (!(0, import_http_client.isHttpClientResponseError)(e)) {
        throw new import_shared_errors2.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await (0, import_shared_client_utils.apiGatewayErrorsHandler)(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ErrorResponseError(
              `400`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(
                errorBody,
                "if permission list is empty, contains unsupported entries or unsupported combinations of entries"
              ),
              e
            );
          } catch (err) {
            if (err instanceof ErrorResponseError) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.putAppSettingsPermissionByObjectIdAndAccessorId:400`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ErrorResponseError(
              `404`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(
                errorBody,
                "No object available for the given objectId or the accessor doesn't have any permissions on this object"
              ),
              e
            );
          } catch (err) {
            if (err instanceof ErrorResponseError) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.putAppSettingsPermissionByObjectIdAndAccessorId:404`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 429: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new TooManyRequests(
              `429`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. Too many requests."),
              e
            );
          } catch (err) {
            if (err instanceof TooManyRequests) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.putAppSettingsPermissionByObjectIdAndAccessorId:429`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 503: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ServiceUnavailable(
              `503`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. Service unavailable."),
              e
            );
          } catch (err) {
            if (err instanceof ServiceUnavailable) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.putAppSettingsPermissionByObjectIdAndAccessorId:503`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        default: {
          if (response.status >= 400 && response.status <= 499) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
              throw new GeneralError(
                `${response.status}`,
                response,
                errorBody,
                (0, import_shared_client_utils.getErrorMessage)(errorBody, "Error."),
                e
              );
            } catch (err) {
              if (err instanceof GeneralError) {
                throw err;
              }
              throw new import_shared_errors2.InvalidResponseError(
                `AppSettingsObjectsClient.putAppSettingsPermissionByObjectIdAndAccessorId:${response.status}`,
                err,
                responseValue,
                "ErrorResponse",
                void 0,
                response
              );
            }
          } else if (response.status >= 500 && response.status <= 599) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
              throw new GeneralError(
                `${response.status}`,
                response,
                errorBody,
                (0, import_shared_client_utils.getErrorMessage)(errorBody, "Error."),
                e
              );
            } catch (err) {
              if (err instanceof GeneralError) {
                throw err;
              }
              throw new import_shared_errors2.InvalidResponseError(
                `AppSettingsObjectsClient.putAppSettingsPermissionByObjectIdAndAccessorId:${response.status}`,
                err,
                responseValue,
                "ErrorResponse",
                void 0,
                response
              );
            }
          } else {
            try {
              const responseValue = await response.body("json");
              throw new import_shared_errors2.ClientRequestError(
                `${response.status}`,
                response,
                responseValue,
                (0, import_shared_client_utils.getErrorMessage)(
                  responseValue,
                  `Unexpected api response: code=${response.status} body="${responseValue}"`
                )
              );
            } catch (error) {
              if (error instanceof import_shared_errors2.ClientRequestError) {
                throw error;
              }
              const responseValue = await response.body("text").catch(() => "");
              throw new import_shared_errors2.InvalidResponseError(
                `${response.status}`,
                error,
                responseValue,
                "json",
                "Unexpected api response",
                response
              );
            }
          }
        }
      }
    }
  }
  /**
   * **Required scope:** app-settings:objects:write<br/>**Required permission:** app-settings:objects:write<br/><br/>Remove the permissions for an existing accessor on this object, anyone with read/write permissions is allowed to delete permissions
   *
   * @example <caption>Code example</caption>
   * import { appSettingsObjectsClient } from "@dynatrace-sdk/client-app-settings-v2";
   *
   * const data =
   *   await appSettingsObjectsClient.deleteAppSettingsPermissionByObjectIdAndAccessorId(
   *     {
   *       objectId: "...",
   *       accessorType: "user",
   *       accessorId: "...",
   *     },
   *   );
   *
   * @throws {ErrorResponseError} No object available for the given objectId or the accessor doesn't have any permissions on this object
   * @throws {TooManyRequests} Failed. Too many requests.
   * @throws {ServiceUnavailable} Failed. Service unavailable.
   * @throws {GeneralError} Error.
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async deleteAppSettingsPermissionByObjectIdAndAccessorId(config) {
    if (!config) {
      throw new import_shared_errors2.ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const query = (0, import_shared_client_utils.toQueryString)({ "admin-access": config.adminAccess });
    try {
      await this.httpClient.send({
        url: `/platform/app-settings/v2/objects/${config.objectId}/permissions/${config.accessorType}/${config.accessorId}${query}`,
        method: "DELETE",
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [204].includes(status);
        }
      });
      return;
    } catch (e) {
      if ((0, import_http_client.isHttpClientAbortError)(e)) {
        throw e;
      }
      if (!(0, import_http_client.isHttpClientResponseError)(e)) {
        throw new import_shared_errors2.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await (0, import_shared_client_utils.apiGatewayErrorsHandler)(response);
      switch (response.status) {
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ErrorResponseError(
              `404`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(
                errorBody,
                "No object available for the given objectId or the accessor doesn't have any permissions on this object"
              ),
              e
            );
          } catch (err) {
            if (err instanceof ErrorResponseError) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.deleteAppSettingsPermissionByObjectIdAndAccessorId:404`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 429: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new TooManyRequests(
              `429`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. Too many requests."),
              e
            );
          } catch (err) {
            if (err instanceof TooManyRequests) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.deleteAppSettingsPermissionByObjectIdAndAccessorId:429`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 503: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ServiceUnavailable(
              `503`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. Service unavailable."),
              e
            );
          } catch (err) {
            if (err instanceof ServiceUnavailable) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.deleteAppSettingsPermissionByObjectIdAndAccessorId:503`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        default: {
          if (response.status >= 400 && response.status <= 499) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
              throw new GeneralError(
                `${response.status}`,
                response,
                errorBody,
                (0, import_shared_client_utils.getErrorMessage)(errorBody, "Error."),
                e
              );
            } catch (err) {
              if (err instanceof GeneralError) {
                throw err;
              }
              throw new import_shared_errors2.InvalidResponseError(
                `AppSettingsObjectsClient.deleteAppSettingsPermissionByObjectIdAndAccessorId:${response.status}`,
                err,
                responseValue,
                "ErrorResponse",
                void 0,
                response
              );
            }
          } else if (response.status >= 500 && response.status <= 599) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
              throw new GeneralError(
                `${response.status}`,
                response,
                errorBody,
                (0, import_shared_client_utils.getErrorMessage)(errorBody, "Error."),
                e
              );
            } catch (err) {
              if (err instanceof GeneralError) {
                throw err;
              }
              throw new import_shared_errors2.InvalidResponseError(
                `AppSettingsObjectsClient.deleteAppSettingsPermissionByObjectIdAndAccessorId:${response.status}`,
                err,
                responseValue,
                "ErrorResponse",
                void 0,
                response
              );
            }
          } else {
            try {
              const responseValue = await response.body("json");
              throw new import_shared_errors2.ClientRequestError(
                `${response.status}`,
                response,
                responseValue,
                (0, import_shared_client_utils.getErrorMessage)(
                  responseValue,
                  `Unexpected api response: code=${response.status} body="${responseValue}"`
                )
              );
            } catch (error) {
              if (error instanceof import_shared_errors2.ClientRequestError) {
                throw error;
              }
              const responseValue = await response.body("text").catch(() => "");
              throw new import_shared_errors2.InvalidResponseError(
                `${response.status}`,
                error,
                responseValue,
                "json",
                "Unexpected api response",
                response
              );
            }
          }
        }
      }
    }
  }
  /**
   * **Required scope:** app-settings:objects:read<br/>**Required permission:** app-settings:objects:read<br/><br/>Get current permissions of the all-users accessor on this object
   *
   * @returns {Promise<AppSettingsAccessorPermissions>} Success
   *
   * @example <caption>Code example</caption>
   * import { appSettingsObjectsClient } from "@dynatrace-sdk/client-app-settings-v2";
   *
   * const data =
   *   await appSettingsObjectsClient.getAppSettingsAllUsersPermissionByObjectId(
   *     { objectId: "..." },
   *   );
   *
   * @throws {ErrorResponseError} No object available for the given objectId or the all-users accessor doesn't have any permissions on this object
   * @throws {TooManyRequests} Failed. Too many requests.
   * @throws {ServiceUnavailable} Failed. Service unavailable.
   * @throws {GeneralError} Error.
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async getAppSettingsAllUsersPermissionByObjectId(config) {
    if (!config) {
      throw new import_shared_errors2.ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const query = (0, import_shared_client_utils.toQueryString)({ "admin-access": config.adminAccess });
    try {
      const response = await this.httpClient.send({
        url: `/platform/app-settings/v2/objects/${config.objectId}/permissions/all-users${query}`,
        method: "GET",
        headers: {
          Accept: "application/json; charset=utf-8, application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
      } catch (err) {
        throw new import_shared_errors2.InvalidResponseError(
          `AppSettingsObjectsClient.getAppSettingsAllUsersPermissionByObjectId:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if ((0, import_http_client.isHttpClientAbortError)(e) || (0, import_shared_errors2.isInvalidResponseError)(e)) {
        throw e;
      }
      if (!(0, import_http_client.isHttpClientResponseError)(e)) {
        throw new import_shared_errors2.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await (0, import_shared_client_utils.apiGatewayErrorsHandler)(response);
      switch (response.status) {
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ErrorResponseError(
              `404`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(
                errorBody,
                "No object available for the given objectId or the all-users accessor doesn't have any permissions on this object"
              ),
              e
            );
          } catch (err) {
            if (err instanceof ErrorResponseError) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.getAppSettingsAllUsersPermissionByObjectId:404`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 429: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new TooManyRequests(
              `429`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. Too many requests."),
              e
            );
          } catch (err) {
            if (err instanceof TooManyRequests) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.getAppSettingsAllUsersPermissionByObjectId:429`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 503: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ServiceUnavailable(
              `503`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. Service unavailable."),
              e
            );
          } catch (err) {
            if (err instanceof ServiceUnavailable) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.getAppSettingsAllUsersPermissionByObjectId:503`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        default: {
          if (response.status >= 400 && response.status <= 499) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
              throw new GeneralError(
                `${response.status}`,
                response,
                errorBody,
                (0, import_shared_client_utils.getErrorMessage)(errorBody, "Error."),
                e
              );
            } catch (err) {
              if (err instanceof GeneralError) {
                throw err;
              }
              throw new import_shared_errors2.InvalidResponseError(
                `AppSettingsObjectsClient.getAppSettingsAllUsersPermissionByObjectId:${response.status}`,
                err,
                responseValue,
                "ErrorResponse",
                void 0,
                response
              );
            }
          } else if (response.status >= 500 && response.status <= 599) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
              throw new GeneralError(
                `${response.status}`,
                response,
                errorBody,
                (0, import_shared_client_utils.getErrorMessage)(errorBody, "Error."),
                e
              );
            } catch (err) {
              if (err instanceof GeneralError) {
                throw err;
              }
              throw new import_shared_errors2.InvalidResponseError(
                `AppSettingsObjectsClient.getAppSettingsAllUsersPermissionByObjectId:${response.status}`,
                err,
                responseValue,
                "ErrorResponse",
                void 0,
                response
              );
            }
          } else {
            try {
              const responseValue = await response.body("json");
              throw new import_shared_errors2.ClientRequestError(
                `${response.status}`,
                response,
                responseValue,
                (0, import_shared_client_utils.getErrorMessage)(
                  responseValue,
                  `Unexpected api response: code=${response.status} body="${responseValue}"`
                )
              );
            } catch (error) {
              if (error instanceof import_shared_errors2.ClientRequestError) {
                throw error;
              }
              const responseValue = await response.body("text").catch(() => "");
              throw new import_shared_errors2.InvalidResponseError(
                `${response.status}`,
                error,
                responseValue,
                "json",
                "Unexpected api response",
                response
              );
            }
          }
        }
      }
    }
  }
  /**
   * **Required scope:** app-settings:objects:write<br/>**Required permission:** app-settings:objects:write<br/><br/>Update the permissions for an existing all-users accessor on this object, anyone with read/write permissions is allowed to update permissions
   *
   * @example <caption>Code example</caption>
   * import { appSettingsObjectsClient } from "@dynatrace-sdk/client-app-settings-v2";
   *
   * const data =
   *   await appSettingsObjectsClient.putAppSettingsAllUsersPermissionByObjectId(
   *     { objectId: "...", body: { permissions: {} } },
   *   );
   *
   * @throws {ErrorResponseError} if permission list is empty, contains unsupported entries or unsupported combinations of entries | No object available for the given objectId or the all-users accessor doesn't have any permissions on this object
   * @throws {TooManyRequests} Failed. Too many requests.
   * @throws {ServiceUnavailable} Failed. Service unavailable.
   * @throws {GeneralError} Error.
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async putAppSettingsAllUsersPermissionByObjectId(config) {
    if (!config) {
      throw new import_shared_errors2.ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const encodedBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformRequest)(config.body, []) : config.body;
    const query = (0, import_shared_client_utils.toQueryString)({ "admin-access": config.adminAccess });
    try {
      await this.httpClient.send({
        url: `/platform/app-settings/v2/objects/${config.objectId}/permissions/all-users${query}`,
        method: "PUT",
        requestBodyType: "json",
        body: encodedBody,
        headers: {
          "Content-Type": "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      return;
    } catch (e) {
      if ((0, import_http_client.isHttpClientAbortError)(e)) {
        throw e;
      }
      if (!(0, import_http_client.isHttpClientResponseError)(e)) {
        throw new import_shared_errors2.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await (0, import_shared_client_utils.apiGatewayErrorsHandler)(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ErrorResponseError(
              `400`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(
                errorBody,
                "if permission list is empty, contains unsupported entries or unsupported combinations of entries"
              ),
              e
            );
          } catch (err) {
            if (err instanceof ErrorResponseError) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.putAppSettingsAllUsersPermissionByObjectId:400`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ErrorResponseError(
              `404`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(
                errorBody,
                "No object available for the given objectId or the all-users accessor doesn't have any permissions on this object"
              ),
              e
            );
          } catch (err) {
            if (err instanceof ErrorResponseError) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.putAppSettingsAllUsersPermissionByObjectId:404`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 429: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new TooManyRequests(
              `429`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. Too many requests."),
              e
            );
          } catch (err) {
            if (err instanceof TooManyRequests) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.putAppSettingsAllUsersPermissionByObjectId:429`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 503: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ServiceUnavailable(
              `503`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. Service unavailable."),
              e
            );
          } catch (err) {
            if (err instanceof ServiceUnavailable) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.putAppSettingsAllUsersPermissionByObjectId:503`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        default: {
          if (response.status >= 400 && response.status <= 499) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
              throw new GeneralError(
                `${response.status}`,
                response,
                errorBody,
                (0, import_shared_client_utils.getErrorMessage)(errorBody, "Error."),
                e
              );
            } catch (err) {
              if (err instanceof GeneralError) {
                throw err;
              }
              throw new import_shared_errors2.InvalidResponseError(
                `AppSettingsObjectsClient.putAppSettingsAllUsersPermissionByObjectId:${response.status}`,
                err,
                responseValue,
                "ErrorResponse",
                void 0,
                response
              );
            }
          } else if (response.status >= 500 && response.status <= 599) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
              throw new GeneralError(
                `${response.status}`,
                response,
                errorBody,
                (0, import_shared_client_utils.getErrorMessage)(errorBody, "Error."),
                e
              );
            } catch (err) {
              if (err instanceof GeneralError) {
                throw err;
              }
              throw new import_shared_errors2.InvalidResponseError(
                `AppSettingsObjectsClient.putAppSettingsAllUsersPermissionByObjectId:${response.status}`,
                err,
                responseValue,
                "ErrorResponse",
                void 0,
                response
              );
            }
          } else {
            try {
              const responseValue = await response.body("json");
              throw new import_shared_errors2.ClientRequestError(
                `${response.status}`,
                response,
                responseValue,
                (0, import_shared_client_utils.getErrorMessage)(
                  responseValue,
                  `Unexpected api response: code=${response.status} body="${responseValue}"`
                )
              );
            } catch (error) {
              if (error instanceof import_shared_errors2.ClientRequestError) {
                throw error;
              }
              const responseValue = await response.body("text").catch(() => "");
              throw new import_shared_errors2.InvalidResponseError(
                `${response.status}`,
                error,
                responseValue,
                "json",
                "Unexpected api response",
                response
              );
            }
          }
        }
      }
    }
  }
  /**
   * **Required scope:** app-settings:objects:write<br/>**Required permission:** app-settings:objects:write<br/><br/>Remove the permissions for an existing all-users accessor on this object, anyone with read/write permissions is allowed to delete permissions
   *
   * @example <caption>Code example</caption>
   * import { appSettingsObjectsClient } from "@dynatrace-sdk/client-app-settings-v2";
   *
   * const data =
   *   await appSettingsObjectsClient.deleteAppSettingsAllUsersPermissionByObjectId(
   *     { objectId: "..." },
   *   );
   *
   * @throws {ErrorResponseError} No object available for the given objectId or the all-users accessor doesn't have any permissions on this object
   * @throws {TooManyRequests} Failed. Too many requests.
   * @throws {ServiceUnavailable} Failed. Service unavailable.
   * @throws {GeneralError} Error.
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async deleteAppSettingsAllUsersPermissionByObjectId(config) {
    if (!config) {
      throw new import_shared_errors2.ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const query = (0, import_shared_client_utils.toQueryString)({ "admin-access": config.adminAccess });
    try {
      await this.httpClient.send({
        url: `/platform/app-settings/v2/objects/${config.objectId}/permissions/all-users${query}`,
        method: "DELETE",
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [204].includes(status);
        }
      });
      return;
    } catch (e) {
      if ((0, import_http_client.isHttpClientAbortError)(e)) {
        throw e;
      }
      if (!(0, import_http_client.isHttpClientResponseError)(e)) {
        throw new import_shared_errors2.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await (0, import_shared_client_utils.apiGatewayErrorsHandler)(response);
      switch (response.status) {
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ErrorResponseError(
              `404`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(
                errorBody,
                "No object available for the given objectId or the all-users accessor doesn't have any permissions on this object"
              ),
              e
            );
          } catch (err) {
            if (err instanceof ErrorResponseError) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.deleteAppSettingsAllUsersPermissionByObjectId:404`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 429: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new TooManyRequests(
              `429`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. Too many requests."),
              e
            );
          } catch (err) {
            if (err instanceof TooManyRequests) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.deleteAppSettingsAllUsersPermissionByObjectId:429`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 503: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ServiceUnavailable(
              `503`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. Service unavailable."),
              e
            );
          } catch (err) {
            if (err instanceof ServiceUnavailable) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.deleteAppSettingsAllUsersPermissionByObjectId:503`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        default: {
          if (response.status >= 400 && response.status <= 499) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
              throw new GeneralError(
                `${response.status}`,
                response,
                errorBody,
                (0, import_shared_client_utils.getErrorMessage)(errorBody, "Error."),
                e
              );
            } catch (err) {
              if (err instanceof GeneralError) {
                throw err;
              }
              throw new import_shared_errors2.InvalidResponseError(
                `AppSettingsObjectsClient.deleteAppSettingsAllUsersPermissionByObjectId:${response.status}`,
                err,
                responseValue,
                "ErrorResponse",
                void 0,
                response
              );
            }
          } else if (response.status >= 500 && response.status <= 599) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
              throw new GeneralError(
                `${response.status}`,
                response,
                errorBody,
                (0, import_shared_client_utils.getErrorMessage)(errorBody, "Error."),
                e
              );
            } catch (err) {
              if (err instanceof GeneralError) {
                throw err;
              }
              throw new import_shared_errors2.InvalidResponseError(
                `AppSettingsObjectsClient.deleteAppSettingsAllUsersPermissionByObjectId:${response.status}`,
                err,
                responseValue,
                "ErrorResponse",
                void 0,
                response
              );
            }
          } else {
            try {
              const responseValue = await response.body("json");
              throw new import_shared_errors2.ClientRequestError(
                `${response.status}`,
                response,
                responseValue,
                (0, import_shared_client_utils.getErrorMessage)(
                  responseValue,
                  `Unexpected api response: code=${response.status} body="${responseValue}"`
                )
              );
            } catch (error) {
              if (error instanceof import_shared_errors2.ClientRequestError) {
                throw error;
              }
              const responseValue = await response.body("text").catch(() => "");
              throw new import_shared_errors2.InvalidResponseError(
                `${response.status}`,
                error,
                responseValue,
                "json",
                "Unexpected api response",
                response
              );
            }
          }
        }
      }
    }
  }
  /**
   * Lists persisted settings objects
   *
   * **Required scope:** app-settings:objects:read<br/>**Required permission:** app-settings:objects:read<br/><br/>Lists persisted settings objects for selected schemas.
   *
   * If nothing is persisted or if all persisted settings objects are not accessible due to missing permissions, no items will be returned.
   *
   * To query the effective values (including schema defaults) please see [getEffectiveAppSettingsValues](#geteffectiveappsettingsvalues).
   *
   * Properties of type **secret** will be included in plain text if the call originates from a serverless function of your app; they will have irreversibly masked values otherwise. This protects these secrets from leaking to users of your app or other third parties.
   *
   * @returns {Promise<AppSettingsObjectsList>} Success. Uses chunked encoding.
   *
   *  Even if a response returns a successful response code it is possible that the result is incomplete due to an internal error.
   *
   *  In this case an 'error' property with information about the problem is added. The caller may decide to work with the incomplete result or do a retry of the operation.
   *
   * @example <caption>Code example</caption>
   * import { appSettingsObjectsClient } from "@dynatrace-sdk/client-app-settings-v2";
   *
   * const data =
   *   await appSettingsObjectsClient.getAppSettingsObjects();
   *
   * @throws {ErrorResponseError} Failed. The input is invalid. | Failed. Forbidden. | Failed. The specified schema was not found.
   * @throws {TooManyRequests} Failed. Too many requests.
   * @throws {ServiceUnavailable} Failed. Service unavailable.
   * @throws {GeneralError} Error.
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async getAppSettingsObjects(config = {}) {
    if (!config) {
      throw new import_shared_errors2.ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const query = (0, import_shared_client_utils.toQueryString)({
      "schema-id": config.schemaId,
      "add-fields": config.addFields,
      filter: config.filter,
      sort: config.sort,
      "page-key": config.pageKey,
      "page-size": config.pageSize,
      "admin-access": config.adminAccess
    });
    try {
      const response = await this.httpClient.send({
        url: `/platform/app-settings/v2/objects${query}`,
        method: "GET",
        headers: {
          Accept: "application/json; charset=utf-8, application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, [
          ["items.modificationInfo.createdTime", { datetime: true }],
          ["items.modificationInfo.lastModifiedTime", { datetime: true }]
        ]) : responseValue;
      } catch (err) {
        throw new import_shared_errors2.InvalidResponseError(
          `AppSettingsObjectsClient.getAppSettingsObjects:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if ((0, import_http_client.isHttpClientAbortError)(e) || (0, import_shared_errors2.isInvalidResponseError)(e)) {
        throw e;
      }
      if (!(0, import_http_client.isHttpClientResponseError)(e)) {
        throw new import_shared_errors2.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await (0, import_shared_client_utils.apiGatewayErrorsHandler)(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ErrorResponseError(
              `400`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. The input is invalid."),
              e
            );
          } catch (err) {
            if (err instanceof ErrorResponseError) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.getAppSettingsObjects:400`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ErrorResponseError(
              `403`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. Forbidden."),
              e
            );
          } catch (err) {
            if (err instanceof ErrorResponseError) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.getAppSettingsObjects:403`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ErrorResponseError(
              `404`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. The specified schema was not found."),
              e
            );
          } catch (err) {
            if (err instanceof ErrorResponseError) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.getAppSettingsObjects:404`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 429: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new TooManyRequests(
              `429`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. Too many requests."),
              e
            );
          } catch (err) {
            if (err instanceof TooManyRequests) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.getAppSettingsObjects:429`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 503: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ServiceUnavailable(
              `503`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. Service unavailable."),
              e
            );
          } catch (err) {
            if (err instanceof ServiceUnavailable) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.getAppSettingsObjects:503`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        default: {
          if (response.status >= 400 && response.status <= 499) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
              throw new GeneralError(
                `${response.status}`,
                response,
                errorBody,
                (0, import_shared_client_utils.getErrorMessage)(errorBody, "Error."),
                e
              );
            } catch (err) {
              if (err instanceof GeneralError) {
                throw err;
              }
              throw new import_shared_errors2.InvalidResponseError(
                `AppSettingsObjectsClient.getAppSettingsObjects:${response.status}`,
                err,
                responseValue,
                "ErrorResponse",
                void 0,
                response
              );
            }
          } else if (response.status >= 500 && response.status <= 599) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
              throw new GeneralError(
                `${response.status}`,
                response,
                errorBody,
                (0, import_shared_client_utils.getErrorMessage)(errorBody, "Error."),
                e
              );
            } catch (err) {
              if (err instanceof GeneralError) {
                throw err;
              }
              throw new import_shared_errors2.InvalidResponseError(
                `AppSettingsObjectsClient.getAppSettingsObjects:${response.status}`,
                err,
                responseValue,
                "ErrorResponse",
                void 0,
                response
              );
            }
          } else {
            try {
              const responseValue = await response.body("json");
              throw new import_shared_errors2.ClientRequestError(
                `${response.status}`,
                response,
                responseValue,
                (0, import_shared_client_utils.getErrorMessage)(
                  responseValue,
                  `Unexpected api response: code=${response.status} body="${responseValue}"`
                )
              );
            } catch (error) {
              if (error instanceof import_shared_errors2.ClientRequestError) {
                throw error;
              }
              const responseValue = await response.body("text").catch(() => "");
              throw new import_shared_errors2.InvalidResponseError(
                `${response.status}`,
                error,
                responseValue,
                "json",
                "Unexpected api response",
                response
              );
            }
          }
        }
      }
    }
  }
  /**
   * Creates a new settings object
   *
   * **Required scope:** app-settings:objects:write<br/>**Required permission:** app-settings:objects:write<br/><br/>Creates a new settings object.
   *
   * @returns {Promise<AppSettingsObjectResponse | void>} Created
   *
   * @example <caption>Code example</caption>
   * import { appSettingsObjectsClient } from "@dynatrace-sdk/client-app-settings-v2";
   *
   * const data =
   *   await appSettingsObjectsClient.postAppSettingsObject({
   *     body: { schemaId: "jira-connection", value: {} },
   *   });
   *
   * @throws {ErrorResponseError} Failed. The input is invalid. | Failed. Forbidden. | Failed. The requested resource doesn't exist.
   * @throws {ConflictingResource} Failed. Conflicting resource.
   * @throws {TooManyRequests} Failed. Too many requests.
   * @throws {ServiceUnavailable} Failed. Service unavailable.
   * @throws {GeneralError} Error.
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async postAppSettingsObject(config) {
    if (!config) {
      throw new import_shared_errors2.ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const encodedBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformRequest)(config.body, []) : config.body;
    const query = (0, import_shared_client_utils.toQueryString)({ "validate-only": config.validateOnly });
    try {
      const response = await this.httpClient.send({
        url: `/platform/app-settings/v2/objects${query}`,
        method: "POST",
        requestBodyType: "json",
        body: encodedBody,
        headers: {
          "Content-Type": "application/json; charset=utf-8",
          Accept: "application/json; charset=utf-8, application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200, 201].includes(status);
        }
      });
      switch (response.status) {
        case 200: {
          return;
        }
        case 201: {
          const responseValue = await response.body("json");
          try {
            return this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
          } catch (err) {
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.postAppSettingsObject:${response.status}`,
              err,
              responseValue,
              void 0,
              void 0,
              void 0
            );
          }
        }
        default: {
          try {
            const responseValue = await response.body("json");
            throw new import_shared_errors2.ClientRequestError(
              `${response.status}`,
              response,
              responseValue,
              (0, import_shared_client_utils.getErrorMessage)(
                responseValue,
                `Unexpected api response: code=${response.status} body="${responseValue}"`
              )
            );
          } catch (error) {
            if (error instanceof import_shared_errors2.ClientRequestError) {
              throw error;
            }
            const responseValue = await response.body("text").catch(() => "");
            throw new import_shared_errors2.InvalidResponseError(
              `${response.status}`,
              error,
              responseValue,
              "json",
              "Unexpected api response",
              response
            );
          }
        }
      }
    } catch (e) {
      if ((0, import_http_client.isHttpClientAbortError)(e) || (0, import_shared_errors2.isInvalidResponseError)(e) || (0, import_shared_errors2.isClientRequestError)(e)) {
        throw e;
      }
      if (!(0, import_http_client.isHttpClientResponseError)(e)) {
        throw new import_shared_errors2.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await (0, import_shared_client_utils.apiGatewayErrorsHandler)(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ErrorResponseError(
              `400`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. The input is invalid."),
              e
            );
          } catch (err) {
            if (err instanceof ErrorResponseError) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.postAppSettingsObject:400`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ErrorResponseError(
              `403`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. Forbidden."),
              e
            );
          } catch (err) {
            if (err instanceof ErrorResponseError) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.postAppSettingsObject:403`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ErrorResponseError(
              `404`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. The requested resource doesn't exist."),
              e
            );
          } catch (err) {
            if (err instanceof ErrorResponseError) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.postAppSettingsObject:404`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 409: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ConflictingResource(
              `409`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. Conflicting resource."),
              e
            );
          } catch (err) {
            if (err instanceof ConflictingResource) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.postAppSettingsObject:409`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 429: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new TooManyRequests(
              `429`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. Too many requests."),
              e
            );
          } catch (err) {
            if (err instanceof TooManyRequests) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.postAppSettingsObject:429`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 503: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ServiceUnavailable(
              `503`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. Service unavailable."),
              e
            );
          } catch (err) {
            if (err instanceof ServiceUnavailable) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.postAppSettingsObject:503`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        default: {
          if (response.status >= 400 && response.status <= 499) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
              throw new GeneralError(
                `${response.status}`,
                response,
                errorBody,
                (0, import_shared_client_utils.getErrorMessage)(errorBody, "Error."),
                e
              );
            } catch (err) {
              if (err instanceof GeneralError) {
                throw err;
              }
              throw new import_shared_errors2.InvalidResponseError(
                `AppSettingsObjectsClient.postAppSettingsObject:${response.status}`,
                err,
                responseValue,
                "ErrorResponse",
                void 0,
                response
              );
            }
          } else if (response.status >= 500 && response.status <= 599) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
              throw new GeneralError(
                `${response.status}`,
                response,
                errorBody,
                (0, import_shared_client_utils.getErrorMessage)(errorBody, "Error."),
                e
              );
            } catch (err) {
              if (err instanceof GeneralError) {
                throw err;
              }
              throw new import_shared_errors2.InvalidResponseError(
                `AppSettingsObjectsClient.postAppSettingsObject:${response.status}`,
                err,
                responseValue,
                "ErrorResponse",
                void 0,
                response
              );
            }
          } else {
            try {
              const responseValue = await response.body("json");
              throw new import_shared_errors2.ClientRequestError(
                `${response.status}`,
                response,
                responseValue,
                (0, import_shared_client_utils.getErrorMessage)(
                  responseValue,
                  `Unexpected api response: code=${response.status} body="${responseValue}"`
                )
              );
            } catch (error) {
              if (error instanceof import_shared_errors2.ClientRequestError) {
                throw error;
              }
              const responseValue = await response.body("text").catch(() => "");
              throw new import_shared_errors2.InvalidResponseError(
                `${response.status}`,
                error,
                responseValue,
                "json",
                "Unexpected api response",
                response
              );
            }
          }
        }
      }
    }
  }
  /**
   * Lists effective settings values
   *
   * **Required scope:** app-settings:objects:read<br/>**Required permission:** app-settings:objects:read<br/><br/>Lists effective settings values for selected schemas.
   * If no object is persisted for a schema with **"multiObject": false**, the default value as defined in the schema will be returned.
   *
   * Properties of type **secret** will be included in plain text if the call originates from a serverless function of your app; they will have irreversibly masked values otherwise. This protects these secrets from leaking to users of your app or other third parties.
   *
   * @returns {Promise<EffectiveAppSettingsValuesList>} Success. Uses chunked encoding.
   *
   *  Even if a response returns a successful response code it is possible that the result is incomplete due to an internal error.
   *
   *  In this case an 'error' property with information about the problem is added. The caller may decide to work with the incomplete result or do a retry of the operation.
   *
   * @example <caption>Code example</caption>
   * import { appSettingsObjectsClient } from "@dynatrace-sdk/client-app-settings-v2";
   *
   * const data =
   *   await appSettingsObjectsClient.getEffectiveAppSettingsValues();
   *
   * @throws {ErrorResponseError} Failed. The input is invalid. | Failed. The specified schema is not found.
   * @throws {TooManyRequests} Failed. Too many requests.
   * @throws {ServiceUnavailable} Failed. Service unavailable.
   * @throws {GeneralError} Error.
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async getEffectiveAppSettingsValues(config = {}) {
    if (!config) {
      throw new import_shared_errors2.ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const query = (0, import_shared_client_utils.toQueryString)({
      "schema-id": config.schemaId,
      "add-fields": config.addFields,
      "page-key": config.pageKey,
      "page-size": config.pageSize,
      "admin-access": config.adminAccess
    });
    try {
      const response = await this.httpClient.send({
        url: `/platform/app-settings/v2/effective-values${query}`,
        method: "GET",
        headers: {
          Accept: "application/json; charset=utf-8, application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, [
          ["items.modificationInfo.createdTime", { datetime: true }],
          ["items.modificationInfo.lastModifiedTime", { datetime: true }]
        ]) : responseValue;
      } catch (err) {
        throw new import_shared_errors2.InvalidResponseError(
          `AppSettingsObjectsClient.getEffectiveAppSettingsValues:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if ((0, import_http_client.isHttpClientAbortError)(e) || (0, import_shared_errors2.isInvalidResponseError)(e)) {
        throw e;
      }
      if (!(0, import_http_client.isHttpClientResponseError)(e)) {
        throw new import_shared_errors2.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await (0, import_shared_client_utils.apiGatewayErrorsHandler)(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ErrorResponseError(
              `400`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. The input is invalid."),
              e
            );
          } catch (err) {
            if (err instanceof ErrorResponseError) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.getEffectiveAppSettingsValues:400`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ErrorResponseError(
              `404`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. The specified schema is not found."),
              e
            );
          } catch (err) {
            if (err instanceof ErrorResponseError) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.getEffectiveAppSettingsValues:404`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 429: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new TooManyRequests(
              `429`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. Too many requests."),
              e
            );
          } catch (err) {
            if (err instanceof TooManyRequests) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.getEffectiveAppSettingsValues:429`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 503: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ServiceUnavailable(
              `503`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. Service unavailable."),
              e
            );
          } catch (err) {
            if (err instanceof ServiceUnavailable) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.getEffectiveAppSettingsValues:503`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        default: {
          if (response.status >= 400 && response.status <= 499) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
              throw new GeneralError(
                `${response.status}`,
                response,
                errorBody,
                (0, import_shared_client_utils.getErrorMessage)(errorBody, "Error."),
                e
              );
            } catch (err) {
              if (err instanceof GeneralError) {
                throw err;
              }
              throw new import_shared_errors2.InvalidResponseError(
                `AppSettingsObjectsClient.getEffectiveAppSettingsValues:${response.status}`,
                err,
                responseValue,
                "ErrorResponse",
                void 0,
                response
              );
            }
          } else if (response.status >= 500 && response.status <= 599) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
              throw new GeneralError(
                `${response.status}`,
                response,
                errorBody,
                (0, import_shared_client_utils.getErrorMessage)(errorBody, "Error."),
                e
              );
            } catch (err) {
              if (err instanceof GeneralError) {
                throw err;
              }
              throw new import_shared_errors2.InvalidResponseError(
                `AppSettingsObjectsClient.getEffectiveAppSettingsValues:${response.status}`,
                err,
                responseValue,
                "ErrorResponse",
                void 0,
                response
              );
            }
          } else {
            try {
              const responseValue = await response.body("json");
              throw new import_shared_errors2.ClientRequestError(
                `${response.status}`,
                response,
                responseValue,
                (0, import_shared_client_utils.getErrorMessage)(
                  responseValue,
                  `Unexpected api response: code=${response.status} body="${responseValue}"`
                )
              );
            } catch (error) {
              if (error instanceof import_shared_errors2.ClientRequestError) {
                throw error;
              }
              const responseValue = await response.body("text").catch(() => "");
              throw new import_shared_errors2.InvalidResponseError(
                `${response.status}`,
                error,
                responseValue,
                "json",
                "Unexpected api response",
                response
              );
            }
          }
        }
      }
    }
  }
  /**
   * Get the effective settings permissions of the calling user in the environment
   *
   * **Required scope:** app-settings:objects:read or app-settings:objects:write<br/>**Required permission:** app-settings:objects:read or app-settings:objects:write
   *
   * @returns {Promise<EffectivePermissions>} Success
   *
   * @example <caption>Code example</caption>
   * import { appSettingsObjectsClient } from "@dynatrace-sdk/client-app-settings-v2";
   *
   * const data =
   *   await appSettingsObjectsClient.resolveEffectivePermissions(
   *     {
   *       body: {
   *         permissions: [
   *           {
   *             permission: "app-settings:objects:read",
   *             context: { schemaId: "..." },
   *           },
   *         ],
   *       },
   *     },
   *   );
   *
   * @throws {ErrorResponseError} Failed. The input is invalid. | Failed. Forbidden. No access to any schema.
   * @throws {GeneralError} Error.
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async resolveEffectivePermissions(config) {
    if (!config) {
      throw new import_shared_errors2.ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const encodedBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformRequest)(config.body, []) : config.body;
    try {
      const response = await this.httpClient.send({
        url: `/platform/app-settings/v2/effective-permissions:resolve`,
        method: "POST",
        requestBodyType: "json",
        body: encodedBody,
        headers: {
          "Content-Type": "application/json; charset=utf-8",
          Accept: "application/json; charset=utf-8, application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
      } catch (err) {
        throw new import_shared_errors2.InvalidResponseError(
          `AppSettingsObjectsClient.resolveEffectivePermissions:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if ((0, import_http_client.isHttpClientAbortError)(e) || (0, import_shared_errors2.isInvalidResponseError)(e)) {
        throw e;
      }
      if (!(0, import_http_client.isHttpClientResponseError)(e)) {
        throw new import_shared_errors2.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await (0, import_shared_client_utils.apiGatewayErrorsHandler)(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ErrorResponseError(
              `400`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. The input is invalid."),
              e
            );
          } catch (err) {
            if (err instanceof ErrorResponseError) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.resolveEffectivePermissions:400`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
            throw new ErrorResponseError(
              `403`,
              response,
              errorBody,
              (0, import_shared_client_utils.getErrorMessage)(errorBody, "Failed. Forbidden. No access to any schema."),
              e
            );
          } catch (err) {
            if (err instanceof ErrorResponseError) {
              throw err;
            }
            throw new import_shared_errors2.InvalidResponseError(
              `AppSettingsObjectsClient.resolveEffectivePermissions:403`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        default: {
          if (response.status >= 400 && response.status <= 499) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
              throw new GeneralError(
                `${response.status}`,
                response,
                errorBody,
                (0, import_shared_client_utils.getErrorMessage)(errorBody, "Error."),
                e
              );
            } catch (err) {
              if (err instanceof GeneralError) {
                throw err;
              }
              throw new import_shared_errors2.InvalidResponseError(
                `AppSettingsObjectsClient.resolveEffectivePermissions:${response.status}`,
                err,
                responseValue,
                "ErrorResponse",
                void 0,
                response
              );
            }
          } else if (response.status >= 500 && response.status <= 599) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
              throw new GeneralError(
                `${response.status}`,
                response,
                errorBody,
                (0, import_shared_client_utils.getErrorMessage)(errorBody, "Error."),
                e
              );
            } catch (err) {
              if (err instanceof GeneralError) {
                throw err;
              }
              throw new import_shared_errors2.InvalidResponseError(
                `AppSettingsObjectsClient.resolveEffectivePermissions:${response.status}`,
                err,
                responseValue,
                "ErrorResponse",
                void 0,
                response
              );
            }
          } else {
            try {
              const responseValue = await response.body("json");
              throw new import_shared_errors2.ClientRequestError(
                `${response.status}`,
                response,
                responseValue,
                (0, import_shared_client_utils.getErrorMessage)(
                  responseValue,
                  `Unexpected api response: code=${response.status} body="${responseValue}"`
                )
              );
            } catch (error) {
              if (error instanceof import_shared_errors2.ClientRequestError) {
                throw error;
              }
              const responseValue = await response.body("text").catch(() => "");
              throw new import_shared_errors2.InvalidResponseError(
                `${response.status}`,
                error,
                responseValue,
                "json",
                "Unexpected api response",
                response
              );
            }
          }
        }
      }
    }
  }
};
var appSettingsObjectsClient = /* @__PURE__ */ new AppSettingsObjectsClient(import_http_client.httpClient);

// packages/client/app-settings-v2/src/lib/error-envelopes/index.ts
var import_shared_errors3 = require("@dynatrace-sdk/shared-errors");

// packages/client/app-settings-v2/src/lib/models/app-settings-permissions-list-item.ts
var AppSettingsPermissionsListItem = /* @__PURE__ */ ((AppSettingsPermissionsListItem2) => {
  AppSettingsPermissionsListItem2["R"] = "r";
  AppSettingsPermissionsListItem2["W"] = "w";
  return AppSettingsPermissionsListItem2;
})(AppSettingsPermissionsListItem || {});

// packages/client/app-settings-v2/src/lib/models/effective-permission-granted.ts
var EffectivePermissionGranted = /* @__PURE__ */ ((EffectivePermissionGranted2) => {
  EffectivePermissionGranted2["True"] = "true";
  EffectivePermissionGranted2["False"] = "false";
  EffectivePermissionGranted2["Condition"] = "condition";
  return EffectivePermissionGranted2;
})(EffectivePermissionGranted || {});

// packages/client/app-settings-v2/src/lib/models/get-app-settings-permission-by-object-id-and-accessor-id-path-accessor-type.ts
var GetAppSettingsPermissionByObjectIdAndAccessorIdPathAccessorType = /* @__PURE__ */ ((GetAppSettingsPermissionByObjectIdAndAccessorIdPathAccessorType2) => {
  GetAppSettingsPermissionByObjectIdAndAccessorIdPathAccessorType2["User"] = "user";
  GetAppSettingsPermissionByObjectIdAndAccessorIdPathAccessorType2["Group"] = "group";
  return GetAppSettingsPermissionByObjectIdAndAccessorIdPathAccessorType2;
})(GetAppSettingsPermissionByObjectIdAndAccessorIdPathAccessorType || {});

// packages/client/app-settings-v2/src/lib/models/identity-type.ts
var IdentityType = /* @__PURE__ */ ((IdentityType2) => {
  IdentityType2["User"] = "user";
  IdentityType2["Group"] = "group";
  IdentityType2["AllUsers"] = "all-users";
  return IdentityType2;
})(IdentityType || {});

// packages/client/app-settings-v2/src/lib/models/resource-context-operations-item.ts
var ResourceContextOperationsItem = /* @__PURE__ */ ((ResourceContextOperationsItem2) => {
  ResourceContextOperationsItem2["Read"] = "read";
  ResourceContextOperationsItem2["Write"] = "write";
  ResourceContextOperationsItem2["Delete"] = "delete";
  return ResourceContextOperationsItem2;
})(ResourceContextOperationsItem || {});

// packages/client/app-settings-v2/src/lib/models/single-permission-request-permission.ts
var SinglePermissionRequestPermission = /* @__PURE__ */ ((SinglePermissionRequestPermission2) => {
  SinglePermissionRequestPermission2["AppSettingsObjectsRead"] = "app-settings:objects:read";
  SinglePermissionRequestPermission2["AppSettingsObjectsWrite"] = "app-settings:objects:write";
  SinglePermissionRequestPermission2["AppSettingsObjectsAdmin"] = "app-settings:objects:admin";
  return SinglePermissionRequestPermission2;
})(SinglePermissionRequestPermission || {});
