// @dynatrace/strato-components/esm/charts/timeseries/utils/chart-configuration.js
import { isEmpty, isUndefined, merge } from "lodash-es";
import { isValidPointsDisplay } from "../../core/components/canvas/utils/is-valid-points-display.js";
import {
  DEFAULT_INFINITE_ZOOM,
  DEFAULT_TRUNCATION_MODE
} from "../../core/constants.js";
import { isValidColorRule } from "../../core/slots/ColorRule.js";
import {
  isCustomColorPalette
} from "../../core/types/color-palette.js";
import { isValidCSSColor } from "../../core/utils/getValidColor.js";
import { getInteractions } from "../../core/utils/interactions.js";
import { isValidColorPalette } from "../../core/utils/is-valid-color-palette.js";
import { isValidLegendPosition } from "../../core/utils/is-valid-legend-position.js";
import { isValidSeriesDisplayMode } from "../../core/utils/is-valid-series-display-mode.js";
import { isValidTooltipVariant } from "../../core/utils/is-valid-tooltip-variant.js";
import { isValidValueRepresentation } from "../../core/utils/is-valid-value-representation.js";
import {
  DEFAULT_SCALE_FALLBACK_COLOR,
  defaultColorPalette,
  defaultGapPolicy,
  defaultLegend,
  defaultPointsDisplay,
  defaultSeriesCurve,
  defaultSeriesDisplayMode,
  defaultTimeseriesChartConfig,
  defaultToolbar,
  defaultTooltipVariant,
  defaultValueRepresentation,
  defaultYAxis
} from "../constants.js";
const sanitizeYAxisConfig = (yAxes) => {
  const reversedYAxes = yAxes.reverse();
  const leftAxis = reversedYAxes.find(
    (axis) => axis?.position === "left" || axis?.position === void 0
  );
  const rightAxis = reversedYAxes.find((axis) => axis?.position === "right");
  const sanitizedYAxes = {};
  if (leftAxis) {
    sanitizedYAxes.left = leftAxis;
  }
  if (rightAxis) {
    sanitizedYAxes.right = rightAxis;
  }
  return sanitizedYAxes;
};
function getYAxisConfiguration(slotsYAxis, contextYAxis) {
  const contextConfig = contextYAxis === void 0 ? {} : sanitizeYAxisConfig(contextYAxis);
  const slotsConfig = slotsYAxis === void 0 ? {} : sanitizeYAxisConfig(slotsYAxis);
  const { left, right } = merge({}, contextConfig, slotsConfig);
  if (left === void 0 && right === void 0) {
    return defaultYAxis;
  }
  const result = {};
  if (left) {
    result.left = merge({}, defaultYAxis.left, left);
  }
  if (right) {
    const defaultRight = { ...defaultYAxis.left, position: "right" };
    result.right = merge({}, defaultRight, right);
  }
  return result;
}
function getLegendConfiguration(contextLegend, slotsLegend) {
  const legendConfig = merge(
    {},
    defaultLegend,
    contextLegend,
    slotsLegend
  );
  legendConfig.position = isValidLegendPosition(legendConfig.position) ? legendConfig.position : defaultLegend.position;
  return legendConfig;
}
function getTooltipConfiguration(contextTooltip, slotsTooltip) {
  const tooltip = merge(
    {},
    { variant: defaultTooltipVariant },
    contextTooltip,
    slotsTooltip
  );
  tooltip.variant = isValidTooltipVariant(tooltip.variant) ? tooltip.variant : defaultTooltipVariant;
  if (tooltip.seriesDisplayMode) {
    tooltip.seriesDisplayMode = isValidSeriesDisplayMode(
      tooltip.seriesDisplayMode
    ) ? tooltip.seriesDisplayMode : defaultSeriesDisplayMode;
  } else {
    tooltip.seriesDisplayMode = tooltip.variant === "shared" ? "single-line" : "multi-line";
  }
  return tooltip;
}
function getXAxisConfiguration(contextXAxis, slotsXAxis) {
  return merge({}, contextXAxis, slotsXAxis);
}
function getThresholdsConfiguration(contextThresholds = [], slotsThresholds = []) {
  const allThresholds = contextThresholds.concat(slotsThresholds).map((threshold) => ({
    ...threshold,
    id: crypto.randomUUID(),
    position: !threshold.position ? "left" : threshold.position,
    color: isValidCSSColor(threshold.color) ? threshold.color : DEFAULT_SCALE_FALLBACK_COLOR
  }));
  if (!allThresholds.length) {
    return void 0;
  }
  const thresholdConfig = {};
  thresholdConfig.left = allThresholds.filter(
    (threshold) => threshold.position === "left"
  );
  thresholdConfig.right = allThresholds.filter(
    (threshold) => threshold.position === "right"
  );
  return thresholdConfig;
}
const isUndefinedOrIsEmpty = (value) => {
  return isUndefined(value) && isEmpty(value);
};
const processToolbarConfig = (config) => {
  if (!config) {
    return config;
  }
  const { hidden, ...rest } = config;
  if (hidden === true) {
    return { ...rest, enabled: false };
  }
  return config;
};
function getToolbar(contextToolbar, slotsChartToolbar, slotDownloadCSV, contextDownloadCSV) {
  const hasNoDownloadData = isUndefinedOrIsEmpty(slotDownloadCSV) && isUndefinedOrIsEmpty(contextDownloadCSV);
  if (!contextToolbar && isEmpty(slotsChartToolbar) && hasNoDownloadData) {
    return { ...defaultToolbar };
  }
  const processedContextToolbar = processToolbarConfig(contextToolbar);
  const processedSlotsToolbar = processToolbarConfig(slotsChartToolbar);
  return merge(
    {},
    defaultToolbar,
    processedContextToolbar,
    processedSlotsToolbar,
    {
      downloadData: slotDownloadCSV ?? contextDownloadCSV
    }
  );
}
function getGapPolicyConfiguration(contextGapPolicy, propsGapPolicy) {
  return propsGapPolicy ?? contextGapPolicy ?? defaultGapPolicy;
}
function getPointsDisplayConfiguration(contextPointsDisplay, propsPointsDisplay) {
  const pointsDisplay = propsPointsDisplay ?? contextPointsDisplay ?? defaultPointsDisplay;
  return isValidPointsDisplay(pointsDisplay) ? pointsDisplay : defaultPointsDisplay;
}
function getValueRepresentationConfiguration(contextValueRepresentation, propsValueRepresentation) {
  const valueRepresentation = propsValueRepresentation ?? contextValueRepresentation ?? defaultValueRepresentation;
  return isValidValueRepresentation(valueRepresentation) ? valueRepresentation : defaultValueRepresentation;
}
function getColorPaletteConfiguration(contextColorPalette, propsColorPalette) {
  const colorPalette = propsColorPalette ?? contextColorPalette ?? defaultColorPalette;
  return isCustomColorPalette(colorPalette) || isValidColorPalette(colorPalette) ? colorPalette : defaultTimeseriesChartConfig.colorPalette ?? defaultColorPalette;
}
function getSeriesCurveConfiguration(contextConfig, propsConfig) {
  return propsConfig ?? contextConfig ?? defaultSeriesCurve;
}
function getInfiniteZoomConfiguration(contextConfig, propsConfig) {
  return propsConfig ?? contextConfig ?? DEFAULT_INFINITE_ZOOM;
}
function getInitialZoomConfiguration(contextConfig, propsConfig) {
  return propsConfig ?? contextConfig;
}
function getCurrentZoomConfiguration(contextConfig, propsConfig) {
  return propsConfig ?? contextConfig;
}
function createTimeseriesConfiguration(slots, contextConfig, timeseriesChartConfigProps) {
  const yAxis = getYAxisConfiguration(slots.yAxis, contextConfig.yAxis);
  const truncationMode = timeseriesChartConfigProps?.truncationMode ?? contextConfig.truncationMode ?? DEFAULT_TRUNCATION_MODE;
  const legend = getLegendConfiguration(contextConfig.legend, slots.legend);
  const tooltip = getTooltipConfiguration(contextConfig.tooltip, slots.tooltip);
  const xAxis = getXAxisConfiguration(contextConfig.xAxis, slots.xAxis);
  const configThresholds = contextConfig.thresholdIndicators || [];
  const thresholds = getThresholdsConfiguration(
    configThresholds,
    slots.thresholds
  );
  const toolbar = getToolbar(contextConfig.toolbar, slots.toolbar);
  const propsInteractions = {
    onZoomChange: timeseriesChartConfigProps.onZoomChange
  };
  const hasContextInteractions = contextConfig && (contextConfig.pan || contextConfig.zoom);
  const chartConfigInteractions = hasContextInteractions ? {
    pan: { enabled: contextConfig.pan?.disabled !== true },
    zoom: { enabled: contextConfig.zoom?.disabled !== true },
    zoomX: { enabled: contextConfig.zoom?.disabled !== true }
  } : {};
  const interactions = getInteractions(
    contextConfig.interactions,
    slots.interactions,
    propsInteractions,
    chartConfigInteractions
  );
  const gapPolicy = getGapPolicyConfiguration(
    contextConfig.gapPolicy,
    timeseriesChartConfigProps?.gapPolicy
  );
  const pointsDisplay = getPointsDisplayConfiguration(
    contextConfig.pointsDisplay,
    timeseriesChartConfigProps?.pointsDisplay
  );
  const valueRepresentation = getValueRepresentationConfiguration(
    contextConfig.valueRepresentation,
    timeseriesChartConfigProps?.valueRepresentation
  );
  const colorPalette = getColorPaletteConfiguration(
    contextConfig.colorPalette,
    timeseriesChartConfigProps?.colorPalette
  );
  const curve = getSeriesCurveConfiguration(
    contextConfig.curve,
    timeseriesChartConfigProps?.curve
  );
  const infiniteZoom = getInfiniteZoomConfiguration(
    contextConfig.infiniteZoom,
    timeseriesChartConfigProps.infiniteZoom
  );
  const initialZoom = getInitialZoomConfiguration(
    contextConfig.initialZoom,
    timeseriesChartConfigProps.initialZoom
  );
  const currentZoom = getCurrentZoomConfiguration(
    contextConfig.currentZoom,
    timeseriesChartConfigProps.currentZoom
  );
  const select = timeseriesChartConfigProps.select ?? slots.select;
  const colorRules = (contextConfig.colorRules ?? []).filter(isValidColorRule).concat(slots.colorRules ?? []);
  return {
    gapPolicy,
    pointsDisplay,
    valueRepresentation,
    colorPalette,
    legend,
    tooltip,
    toolbar,
    xAxis,
    yAxis,
    annotations: slots.annotations,
    truncationMode,
    thresholds,
    interactions,
    initialZoom,
    currentZoom,
    infiniteZoom,
    curve,
    select,
    colorRules
  };
}
function parseYAxisConfig(yAxis) {
  const { left, right } = yAxis;
  const yAxisConfig = [];
  if (left) {
    yAxisConfig.push(left);
  }
  if (right) {
    yAxisConfig.push(right);
  }
  return yAxisConfig;
}
function parseThresholdsConfig(thresholds) {
  const { left, right } = thresholds;
  const thresholdsConfig = [];
  if (left) {
    thresholdsConfig.push(...left);
  }
  if (right) {
    thresholdsConfig.push(...right);
  }
  return thresholdsConfig;
}
function normalizeConfig(config) {
  const { thresholds, ...rest } = config;
  const thresholdIndicators = thresholds ? parseThresholdsConfig(thresholds) : void 0;
  return {
    ...rest,
    yAxis: parseYAxisConfig(config.yAxis),
    thresholdIndicators
  };
}
const dateFormat = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z$/;
function reviver(key, value) {
  if (typeof value === "string" && dateFormat.test(value)) {
    return new Date(value);
  }
  return value;
}
function deserializeConfig(serializedConfig) {
  try {
    return merge(
      {},
      defaultTimeseriesChartConfig,
      JSON.parse(serializedConfig, reviver)
    );
  } catch {
    return defaultTimeseriesChartConfig;
  }
}
export {
  createTimeseriesConfiguration,
  deserializeConfig,
  getColorPaletteConfiguration,
  getCurrentZoomConfiguration,
  getGapPolicyConfiguration,
  getInfiniteZoomConfiguration,
  getInitialZoomConfiguration,
  getLegendConfiguration,
  getPointsDisplayConfiguration,
  getSeriesCurveConfiguration,
  getThresholdsConfiguration,
  getToolbar,
  getTooltipConfiguration,
  getValueRepresentationConfiguration,
  getXAxisConfiguration,
  getYAxisConfiguration,
  normalizeConfig
};
