// @dynatrace/strato-components/esm/charts/timeseries/utils/get-datapoint-value.js
import { getDatapointTime } from "./timeseries.js";
import { getRelativeValue } from "../../core/utils/get-relative-value.js";
import { getUnitAsString } from "../../core/utils/get-unit-as-string.js";
import { safeGet } from "../../core/utils/safe-get.js";
const getDatapointKey = (timestamp, unit) => `${timestamp}_${unit}`;
function getDatapointValue(datapoint, variant, unit, relativeGroupedValues) {
  if (variant === "band") {
    const { y0, y1 } = datapoint;
    if (relativeGroupedValues) {
      const seriesUnit = getUnitAsString(unit);
      const { maxValuesByUnit } = relativeGroupedValues;
      const total = safeGet(maxValuesByUnit, seriesUnit);
      const relativeY0Value = getRelativeValue(y0, total);
      const relativeY1Value = getRelativeValue(y1, total);
      return {
        y0: relativeY0Value,
        absoluteY0: y0,
        y1: relativeY1Value,
        absoluteY1: y1
      };
    }
    return { y0, y1 };
  }
  const { value, start, end, center } = datapoint;
  if (relativeGroupedValues) {
    const { groupedValues, maxValuesByUnit } = relativeGroupedValues;
    const seriesUnit = getUnitAsString(unit);
    const datapointKey = getDatapointKey(
      getDatapointTime(start, end, center),
      seriesUnit
    );
    const sumOfSeriesValues = safeGet(groupedValues, datapointKey) ?? 0;
    const isTotalValuePerUnit = relativeGroupedValues.uniqueUnits.get(seriesUnit) || variant === "line";
    const total = isTotalValuePerUnit ? maxValuesByUnit.get(seriesUnit) : sumOfSeriesValues;
    const relativeValue = getRelativeValue(value, total);
    return { absolute: value, value: relativeValue };
  }
  return { value };
}
export {
  getDatapointKey,
  getDatapointValue
};
