// @dynatrace/strato-components/esm/charts/timeseries/utils/iterate-config-slots.js
import { isEmpty, merge } from "lodash-es";
import { Children } from "react";
import { isElement } from "react-is";
import { Intent } from "../../../core/slots/Intent/Intent.js";
import { isSlot } from "../../../core/utils/is-slot.js";
import { populateMissingMarkerProps } from "../../core/components/annotations/utils/populate-missing-marker-props.js";
import { populateMissingTrackProps } from "../../core/components/annotations/utils/populate-missing-track-props.js";
import { ChartInteractions } from "../../core/slots/chart-interactions/ChartInteractions.js";
import { ColorRule, isValidColorRule } from "../../core/slots/ColorRule.js";
import { ChartToolbar } from "../../core/slots/toolbar/ChartToolbar.js";
import { sanitizeProps } from "../../core/slots/toolbar/CompactToolbar.js";
import { DownloadData } from "../../core/slots/toolbar/DownloadData.js";
import { extractBehaviorTrackingProps } from "../../core/utils/extract-behavior-tracking-props.js";
import { getConfigPlusDefaultOrEmpty } from "../../core/utils/interactions.js";
import { iterateChartInteractionsSlots } from "../../core/utils/iterate-chart-interactions-slots.js";
import { Annotations } from "../slots/annotations/Annotations.js";
import { Marker } from "../slots/annotations/Marker.js";
import { AnnotationsTooltip } from "../slots/annotations/Tooltip.js";
import { Track } from "../slots/annotations/Track.js";
import { TimeseriesChartDownloadCSVSlot } from "../slots/DownloadCSV.js";
import { Legend } from "../slots/Legend.js";
import { Pan } from "../slots/Pan.js";
import { Select } from "../slots/Select.js";
import { Area } from "../slots/shapes/Area.js";
import { Band } from "../slots/shapes/Band.js";
import { Bar } from "../slots/shapes/Bar.js";
import { Line } from "../slots/shapes/Line.js";
import { ThresholdIndicator } from "../slots/ThresholdIndicator.js";
import { Toolbar } from "../slots/Toolbar.js";
import { Tooltip } from "../slots/Tooltip.js";
import { XAxis } from "../slots/XAxis.js";
import { YAxis } from "../slots/YAxis.js";
import { Zoom } from "../slots/Zoom.js";
const slotInteractionsDefaultValues = {
  zoom: { enabled: true },
  zoomX: { enabled: true },
  pan: { enabled: true }
};
const chartInteractionsDefaultValues = {
  zoom: { enabled: false },
  zoomX: { enabled: false },
  pan: { enabled: false }
};
const getInteractionsConfig = (chartConfig, slotsConfig) => {
  const chartApiValues = getConfigPlusDefaultOrEmpty(
    chartInteractionsDefaultValues,
    chartConfig
  );
  const slotsApiValues = getConfigPlusDefaultOrEmpty(
    slotInteractionsDefaultValues,
    slotsConfig
  );
  return merge({}, chartApiValues, slotsApiValues);
};
const disabledOnlyWhenSpecified = ({ disabled }) => disabled;
const iterateConfigSlots = (children) => {
  let slots = {
    toolbar: {},
    legend: {},
    xAxis: {},
    tooltip: {},
    yAxis: [],
    thresholds: [],
    interactions: {},
    colorRules: []
  };
  let intents;
  let slotToolbar = {};
  let chartToolbar = {};
  let slotInteractions = {};
  let chartInteractions = {};
  Children.forEach(children, (child) => {
    if (isSlot(child, ChartInteractions)) {
      const { children: children2, ...previousInteractions } = child.props;
      const interactions = iterateChartInteractionsSlots(children2);
      chartInteractions = {
        ...previousInteractions,
        ...interactions
      };
    } else if (isSlot(child, ChartToolbar)) {
      const { children: children2, ...toolbar } = child.props;
      const childrenToolbar = iterateChartToolbarSlots(children2);
      chartToolbar = {
        enabled: true,
        ...toolbar,
        ...childrenToolbar
      };
    } else if (isSlot(child, Toolbar)) {
      slotToolbar = {
        enabled: child.props.hidden !== true,
        ...slotToolbar,
        ...sanitizeProps(child.props),
        ...extractBehaviorTrackingProps(child.props)
      };
    } else if (isSlot(child, TimeseriesChartDownloadCSVSlot)) {
      slotToolbar = { ...slotToolbar, downloadData: { enabled: true } };
    } else if (isSlot(child, Zoom)) {
      const enabled = !disabledOnlyWhenSpecified(child.props);
      slotInteractions = {
        ...slotInteractions,
        zoom: { enabled },
        zoomX: { enabled }
      };
    } else if (isSlot(child, Pan)) {
      const enabled = !disabledOnlyWhenSpecified(child.props);
      slotInteractions = { ...slotInteractions, pan: { enabled } };
    } else if (isSlot(child, Legend)) {
      const legend = child.props;
      slots = {
        ...slots,
        legend
      };
    } else if (isSlot(child, Intent)) {
      intents = intents ? [...intents, child.props] : [child.props];
    } else if (isSlot(child, XAxis)) {
      const xAxis = child.props;
      slots = {
        ...slots,
        xAxis
      };
    } else if (isSlot(child, YAxis)) {
      const yAxis = child.props;
      slots.yAxis.push(yAxis);
    } else if (isSlot(child, Tooltip)) {
      const tooltip = child.props;
      slots = {
        ...slots,
        tooltip
      };
    } else if (isSlot(child, Annotations)) {
      const { children: children2, ...annotationsProps } = child.props;
      const { tracks, tooltip } = iterateTrackSlots(children2);
      const annotations = {
        ...annotationsProps,
        tracks,
        tooltip
      };
      slots = {
        ...slots,
        annotations
      };
    } else if (isSlot(child, ThresholdIndicator)) {
      const thresholdProps = child.props;
      slots.thresholds.push(thresholdProps);
    } else if (isSlot(child, Select)) {
      const select = child.props;
      slots = {
        ...slots,
        select
      };
    } else if (isSlot(child, ColorRule)) {
      const colorRule = child.props;
      if (isValidColorRule(colorRule)) {
        slots.colorRules.push(colorRule);
      }
    }
  });
  const hasInteractionConfig = Object.keys(chartInteractions).length > 0 || Object.keys(slotInteractions).length > 0;
  slots.interactions = hasInteractionConfig ? getInteractionsConfig(chartInteractions, slotInteractions) : {};
  const hasToolbarConfig = !isEmpty(slotToolbar) || !isEmpty(chartToolbar) || intents;
  if (hasToolbarConfig) {
    slots.toolbar = {
      ...chartToolbar,
      ...slotToolbar,
      enabled: slotToolbar.enabled ?? chartToolbar.enabled ?? true,
      intents
    };
  }
  return slots;
};
const iterateTrackSlots = (children) => {
  const tracks = [];
  let tooltip = {
    template: void 0,
    hidden: false
  };
  Children.forEach(children, (child) => {
    if (isSlot(child, Track)) {
      const { children: children2, ...trackProps } = child.props;
      const currentTrack = {
        ...trackProps,
        markers: []
      };
      Children.forEach(children2, (child2) => {
        if (isSlot(child2, Marker)) {
          const marker = populateMissingMarkerProps(child2.props, trackProps);
          currentTrack.markers.push(marker);
        }
      });
      const track = populateMissingTrackProps(currentTrack);
      tracks.push(track);
    } else if (isSlot(child, AnnotationsTooltip)) {
      const { children: children2, hidden } = child.props;
      tooltip = {
        template: children2,
        hidden: hidden ?? false
      };
    }
  });
  return { tracks, tooltip };
};
const iterateChartToolbarSlots = (children) => {
  let toolbar = {};
  Children.forEach(children, (childToolbar) => {
    if (isElement(childToolbar) && childToolbar.type === DownloadData) {
      toolbar = {
        ...toolbar,
        downloadData: { enabled: true }
      };
    }
  });
  return toolbar;
};
const iterateChartSeriesSlots = (children) => {
  const line = [];
  const bar = [];
  const area = [];
  const band = [];
  const shapePriority = [];
  Children.forEach(children, (childSlot) => {
    if (isElement(childSlot)) {
      switch (childSlot.type) {
        case Line: {
          const { data, ...overrides } = childSlot.props;
          line.push({ ...data, ...overrides });
          if (!shapePriority.includes("line")) {
            shapePriority.push("line");
          }
          break;
        }
        case Bar: {
          const { data, ...overrides } = childSlot.props;
          bar.push({ ...data, ...overrides });
          if (!shapePriority.includes("bar")) {
            shapePriority.push("bar");
          }
          break;
        }
        case Area: {
          const { data, ...overrides } = childSlot.props;
          area.push({ ...data, ...overrides });
          if (!shapePriority.includes("area")) {
            shapePriority.push("area");
          }
          break;
        }
        case Band: {
          const {
            data,
            strokeOnly = false,
            color,
            gapPolicy
          } = childSlot.props;
          band.push({ ...data, strokeOnly, color, gapPolicy });
          if (!shapePriority.includes("band")) {
            shapePriority.push("band");
          }
          break;
        }
      }
    }
  });
  return { band, line, area, bar, shapePriority };
};
export {
  iterateChartSeriesSlots,
  iterateConfigSlots
};
