// @dynatrace/strato-components/esm/charts/timeseries/utils/timeseries.js
import { iterateChartSeriesSlots } from "./iterate-config-slots.js";
import { TIMESERIES_CHART_DEFAULT_VARIANT } from "../constants.js";
function isEmptyTimeseries(timeseries) {
  return timeseries.length === 0 || timeseries.every((series) => series.datapoints.length === 0);
}
function splitTimeseriesPerVariant(timeseries, children, defaultVariant) {
  const seriesFromSlots = iterateChartSeriesSlots(children);
  const timeseriesPerVariant = {
    line: [],
    area: [],
    bar: [],
    band: [],
    shapePriority: []
  };
  timeseriesPerVariant[defaultVariant ?? TIMESERIES_CHART_DEFAULT_VARIANT] = timeseries;
  const shapePriority = defaultVariant && seriesFromSlots.shapePriority.includes(defaultVariant) ? seriesFromSlots.shapePriority : [...seriesFromSlots.shapePriority, defaultVariant];
  return {
    line: timeseriesPerVariant.line.concat(seriesFromSlots.line),
    area: timeseriesPerVariant.area.concat(seriesFromSlots.area),
    bar: timeseriesPerVariant.bar.concat(seriesFromSlots.bar),
    band: timeseriesPerVariant.band.concat(seriesFromSlots.band),
    shapePriority
  };
}
function getDatapointTime(start, end, center) {
  if (center) {
    return center.getTime();
  }
  if (end) {
    return start.getTime() + (end.getTime() - start.getTime()) / 2;
  }
  return start.getTime();
}
const hasSeriesWithDatapoints = (dataPerVariant) => {
  const data = mergeTimeseriesFromVariants(dataPerVariant);
  return data.length > 0 && data.some((series) => series.datapoints.length);
};
function mergeTimeseriesFromVariants(dataPerVariant) {
  const { line, area, bar, band } = dataPerVariant;
  return line.concat(area, bar, band);
}
function hasMultiGeometries(data) {
  const filteredData = Object.keys(data).map(
    (variant) => variant !== "shapePriority" ? data[variant] : []
  );
  const valuesPerVariant = filteredData.filter(
    (geometryData) => geometryData.length > 0
  );
  return valuesPerVariant.length > 1;
}
export {
  getDatapointTime,
  hasMultiGeometries,
  hasSeriesWithDatapoints,
  isEmptyTimeseries,
  mergeTimeseriesFromVariants,
  splitTimeseriesPerVariant
};
