// @dynatrace/strato-components/esm/charts/timeseries/utils/yAxis.js
import { isUndefined } from "lodash-es";
import { getYAxisValueDomainsByUnit } from "./yScaleDomain.js";
import { buildScale } from "../../core/utils/core-scale.js";
function getYScaleByUnit(seriesUnit, yScales) {
  if (seriesUnit === yScales.left?.unit) {
    return yScales.left.scale;
  }
  const axisScale = yScales.right || yScales.left;
  return axisScale.scale;
}
function getYScaleTypeByUnit(seriesUnit, yScales, axisScaleType) {
  if (seriesUnit === yScales.left?.unit) {
    return axisScaleType.left?.scale ?? "linear";
  }
  return (axisScaleType.right?.scale || axisScaleType.left?.scale) ?? "linear";
}
const getConfigMinMax = (yAxisConfig) => {
  if (!yAxisConfig) {
    return void 0;
  }
  return {
    min: yAxisConfig.min,
    max: yAxisConfig.max
  };
};
const isMissingData = (dataBoundaries) => isUndefined(dataBoundaries) || isUndefined(dataBoundaries.min) || isUndefined(dataBoundaries.max);
const buildYScale = (height, dataBoundaries, configMinMax, scaleType = "linear") => {
  return {
    scale: buildScale(scaleType, [height, 0], dataBoundaries, configMinMax),
    unit: dataBoundaries.unit
  };
};
const parseMapToArray = (mapBoundariesUnit) => {
  const response = [];
  for (const [unit, { min, max }] of mapBoundariesUnit) {
    response.push({ unit, min, max });
  }
  return response;
};
const getDataBoundaries = (timeseries, seriesMap, datapointsMap) => {
  const leftMetadata = {};
  const rightMetadata = {};
  const mapBoundariesUnit = timeseries.shapePriority.reduce(
    (boundariesByUnit, variant) => {
      const shapeDomains = getYAxisValueDomainsByUnit(
        timeseries[variant],
        variant,
        leftMetadata,
        rightMetadata,
        seriesMap,
        datapointsMap
      );
      shapeDomains.forEach(({ min, max, unit }) => {
        if (boundariesByUnit.has(unit)) {
          const current = boundariesByUnit.get(unit);
          const boundaries = {
            min: min ? Math.min(current.min, min) : current.min,
            max: max ? Math.max(current.max, max) : current.max
          };
          boundariesByUnit.set(unit, boundaries);
        } else {
          boundariesByUnit.set(unit, { min, max });
        }
      });
      return boundariesByUnit;
    },
    /* @__PURE__ */ new Map()
  );
  return parseMapToArray(mapBoundariesUnit);
};
const assignBoundariesToYAxes = (leftConfigMinMax, rightConfigMinMax, firstUnitBoundaries, secondUnitBoundaries) => {
  const hasRightAxis = !isUndefined(rightConfigMinMax);
  const hasSecondDomain = !isUndefined(secondUnitBoundaries);
  if (!hasSecondDomain) {
    return hasRightAxis ? { rightUnitBoundaries: firstUnitBoundaries } : { leftUnitBoundaries: firstUnitBoundaries };
  }
  return {
    leftUnitBoundaries: firstUnitBoundaries,
    rightUnitBoundaries: secondUnitBoundaries
  };
};
const shouldReturnUserConfig = (mainConfig) => !isUndefined(mainConfig);
function shouldBuildScale(unitBoundaries, configMinMax) {
  return !isMissingData(unitBoundaries) && shouldReturnUserConfig(configMinMax);
}
function shouldDisplayAxis(unitBoundaries, defaultDisplay, configMinMax) {
  return unitBoundaries && (shouldBuildScale(unitBoundaries, configMinMax) || defaultDisplay);
}
const buildYScales = (timeseries, config, height, seriesMap, datapointsMap) => {
  const leftScaleType = config.left?.scale;
  const rightScaleType = config.right?.scale;
  const leftConfigMinMax = getConfigMinMax(config.left);
  const rightConfigMinMax = getConfigMinMax(config.right);
  const [firstUnitBoundaries, secondUnitBoundaries] = getDataBoundaries(
    timeseries,
    seriesMap,
    datapointsMap
  );
  const { leftUnitBoundaries, rightUnitBoundaries } = assignBoundariesToYAxes(
    leftConfigMinMax,
    rightConfigMinMax,
    firstUnitBoundaries,
    secondUnitBoundaries
  );
  const displayBothAxisByDefault = !leftConfigMinMax && !rightConfigMinMax;
  const displayLeftAxis = shouldDisplayAxis(
    leftUnitBoundaries,
    displayBothAxisByDefault,
    leftConfigMinMax
  );
  const displayRightAxis = shouldDisplayAxis(
    rightUnitBoundaries,
    displayBothAxisByDefault,
    rightConfigMinMax
  );
  const left = displayLeftAxis ? buildYScale(
    height,
    leftUnitBoundaries,
    leftConfigMinMax,
    leftScaleType
  ) : void 0;
  const right = displayRightAxis ? buildYScale(
    height,
    rightUnitBoundaries,
    rightConfigMinMax,
    rightScaleType
  ) : void 0;
  return {
    left,
    right
  };
};
export {
  assignBoundariesToYAxes,
  buildYScales,
  getYScaleByUnit,
  getYScaleTypeByUnit
};
