// @dynatrace/strato-components/esm/charts/timeseries/utils/yScaleDomain.js
import { extent } from "d3-array";
import { safeGet } from "../../core/utils/safe-get.js";
const getValueOrGap = (value, isGap) => isGap ? void 0 : value ?? 0;
const getYTimeSeries = (datapoint, datapointsMap) => {
  const metadata = safeGet(datapointsMap, datapoint);
  return getValueOrGap(metadata.value, metadata.isGap);
};
const getYBand0 = (datapoint, datapointsMap) => {
  const metadata = safeGet(datapointsMap, datapoint);
  return getValueOrGap(metadata.y0, metadata.isGap);
};
const getYBand1 = (datapoint, datapointsMap) => {
  const metadata = safeGet(datapointsMap, datapoint);
  return getValueOrGap(metadata.y1, metadata.isGap);
};
const getDomainByUnit = (unit, seriesMap, datapointsMap, data) => {
  const datapoints = data.series.filter((singleSeries) => safeGet(seriesMap, singleSeries).unit === unit).flatMap(
    (currentSeries) => currentSeries.datapoints
  );
  const { min, max } = getMinMax(datapointsMap, {
    variant: data.variant,
    datapoints
  });
  return {
    min,
    max,
    unit
  };
};
function getYAxisValueDomainsByUnit(series, variant, leftAxisMetadata, rightAxisMetadata, seriesMap, datapointsMap) {
  if (series.length === 0) {
    return [];
  }
  const otherGeometriesUnits = [
    leftAxisMetadata.unit,
    rightAxisMetadata.unit
  ].filter((u) => u !== void 0);
  const seriesUnit = series.map((_series) => safeGet(seriesMap, _series).unit);
  const scaleUnits = [.../* @__PURE__ */ new Set([...otherGeometriesUnits, ...seriesUnit])];
  return scaleUnits.map(
    (unit) => getDomainByUnit(unit, seriesMap, datapointsMap, {
      series,
      variant
    })
  );
}
const getLineMinMax = (datapointsMap, datapoints) => {
  const [min = 0, max = 0] = extent(
    datapoints,
    (dp) => getYTimeSeries(dp, datapointsMap)
  );
  return { min, max };
};
const getStackedMinMax = (datapointsMap, datapoints) => {
  const [min = 0, max = 0] = extent(
    datapoints.map((datapoint) => {
      const metadataDatapoint = safeGet(datapointsMap, datapoint);
      if (metadataDatapoint.isGap) {
        return void 0;
      }
      return metadataDatapoint.cumulativeValue ?? 0;
    })
  );
  return { min, max };
};
const getBandMinMax = (datapointsMap, datapoints) => {
  const [minY0 = 0, maxY0 = 0] = extent(
    datapoints,
    (dp) => getYBand0(dp, datapointsMap)
  );
  const [minY1 = 0, maxY1 = 0] = extent(
    datapoints,
    (dp) => getYBand1(dp, datapointsMap)
  );
  const min = Math.min(minY0, minY1);
  const max = Math.max(maxY0, maxY1);
  return { min, max };
};
const MinMaxStrategyMap = {
  band: getBandMinMax,
  line: getLineMinMax,
  area: getStackedMinMax,
  bar: getStackedMinMax
};
const getMinMax = (datapointsMap, data) => {
  return MinMaxStrategyMap[data.variant](
    datapointsMap,
    data.datapoints
  );
};
export {
  getBandMinMax,
  getLineMinMax,
  getStackedMinMax,
  getYAxisValueDomainsByUnit
};
