// @dynatrace/strato-components/esm/charts/meter-bar/shared/components/meter-bar-layout/MeterBarLayout.sty.js
import "./MeterBarLayout.css";
import { createRuntimeFn as _7a468 } from "@vanilla-extract/recipes/createRuntimeFn";
var chartContainer = "_1m3udw19-3-6-0";
var container = _7a468({ defaultClassName: "_1m3udw1a-3-6-0", variantClassNames: { autoSize: { true: "_1m3udw1b-3-6-0", false: "_1m3udw1c-3-6-0" }, showLoading: { true: "_1m3udw1d-3-6-0" } }, defaultVariants: {}, compoundVariants: [] });
var containerWrapper = "_1m3udw1e-3-6-0";
var labelContainer = "_1m3udw18-3-6-0";
var legendContainer = "_1m3udw17-3-6-0";
var maxContainer = "_1m3udw16-3-6-0";
var minContainer = "_1m3udw15-3-6-0";
var toolbarAnimatedContainer = "_1m3udw14-3-6-0";
var valueAnimatedContainer = "_1m3udw13-3-6-0";
var valueContainer = _7a468({ defaultClassName: "_1m3udw10-3-6-0", variantClassNames: { inline: { true: "_1m3udw11-3-6-0", false: "_1m3udw12-3-6-0" } }, defaultVariants: {}, compoundVariants: [] });
export {
  chartContainer,
  container,
  containerWrapper,
  labelContainer,
  legendContainer,
  maxContainer,
  minContainer,
  toolbarAnimatedContainer,
  valueAnimatedContainer,
  valueContainer
};
