// @dynatrace/strato-components/esm/charts/meter-bar/shared/components/meter-bar-layout/MeterBarLayout.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import {
  forwardRef,
  Fragment,
  useImperativeHandle,
  useRef
} from "react";
import { MeterBarLayoutTopRightSlot } from "./components/MeterBarLayoutTopRightSlot.js";
import { useComputedSegmentSize } from "./hooks/useComputedSegmentSize.js";
import {
  chartContainer,
  container,
  containerWrapper,
  labelContainer,
  legendContainer,
  maxContainer,
  minContainer
} from "./MeterBarLayout.sty.js";
import { Chart } from "./slots/Chart.js";
import { Icon } from "./slots/Icon/Icon.js";
import { Label } from "./slots/Label/Label.js";
import { Legend } from "./slots/Legend.js";
import { Max } from "./slots/Max/Max.js";
import { Min } from "./slots/Min/Min.js";
import { Providers } from "./slots/Providers.js";
import { Value } from "./slots/Value/Value.js";
import { extractSlots } from "./utils/extract-slots.js";
import { useBehavioralTrackingProps } from "../../../../../core/hooks/useBehavioralTrackingProps.js";
import { useMergeRefs } from "../../../../../core/hooks/useMergeRefs.js";
import { Flex } from "../../../../../layouts/flex/Flex.js";
import { Grid } from "../../../../../layouts/grid/Grid.js";
import { LoaderOverlay } from "../../../../core/components/loader-overlay/LoaderOverlay.js";
import { useTooltipImperativeHandle } from "../../../../core/components/overlay-tooltip/hooks/useTooltipImperativeHandle.js";
import { coerceSizeValue } from "../../../../core/utils/coerceSizeValue.js";
import {
  DEFAULT_HEIGHT,
  DEFAULT_WIDTH,
  METER_BAR_SIZE
} from "../../constants.js";
import { useMeterBarChartLoading } from "../../context/MeterBarChartLoadingContext.js";
import { useMeterBarChartToolbar } from "../../hooks/useMeterBarChartToolbar.js";
import { consumerLoadingStyle } from "../../styles.sty.js";
import { MeterBarSizeProviders } from "../meter-bar-providers/MeterBarSizeProviders.js";
const MeterBarLayoutRoot = forwardRef((props, forwardedRef) => {
  const {
    width,
    height,
    size,
    children,
    className,
    style,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "data-testid": dataTestId,
    "data-dt-component": dataDtComponent,
    "aria-label": ariaLabel,
    userRef,
    consumerConfig = {}
  } = props;
  const containerRef = useRef(null);
  const { label, min, max, value, chart, icon, legend, providers } = extractSlots(children);
  const [behavioralTrackingProps] = useBehavioralTrackingProps(props);
  const loading = useMeterBarChartLoading();
  const { isToolbarEnabled, modifyHoverState, toolbarRef } = useMeterBarChartToolbar();
  const tooltipActions = useTooltipImperativeHandle();
  const ref = useMergeRefs([containerRef, forwardedRef]);
  useImperativeHandle(userRef, () => {
    return {
      element: containerRef.current,
      getConfig: () => JSON.stringify(consumerConfig),
      ...tooltipActions
    };
  });
  const hasAutoSize = size === METER_BAR_SIZE.Auto;
  const gridTemplateRows = hasAutoSize ? "auto 1fr auto auto" : "auto";
  const {
    segmentSize: { height: segmentHeight, width: segmentWidth },
    labelRef,
    minLabelRef,
    maxLabelRef,
    plotAreaRef,
    legendRef
  } = useComputedSegmentSize();
  const chartWidth = coerceSizeValue(width) ?? DEFAULT_WIDTH;
  const chartHeight = coerceSizeValue(height) ?? DEFAULT_HEIGHT;
  const ExternalProviders = providers ? providers.Component : Fragment;
  const externalProvidersProps = providers ? providers.props : {};
  return /* @__PURE__ */ jsx(
    "div",
    {
      ...behavioralTrackingProps,
      className: clsx(className, containerWrapper),
      style: {
        width: chartWidth,
        height: chartHeight,
        ...style
      },
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "data-dt-component": dataDtComponent,
      ref,
      children: /* @__PURE__ */ jsx(
        "div",
        {
          className: container({ showLoading: loading, autoSize: hasAutoSize }),
          children: /* @__PURE__ */ jsx(
            Grid,
            {
              className: container({ autoSize: hasAutoSize }),
              gridTemplateAreas: "\n            'label label label value'\n            'chart chart chart chart'\n            'min min max max'\n            'legend legend legend legend'",
              gridTemplateColumns: "25% 25% 25% 25%",
              gridTemplateRows,
              rowGap: 0,
              columnGap: 0,
              "aria-label": ariaLabel,
              role: "img",
              onMouseEnter: () => modifyHoverState(true),
              onMouseLeave: () => modifyHoverState(false),
              onFocus: () => modifyHoverState(true),
              onBlur: () => modifyHoverState(false),
              children: /* @__PURE__ */ jsx(
                MeterBarSizeProviders,
                {
                  size,
                  chartWidth: containerRef.current?.clientWidth ?? 0,
                  chartHeight: containerRef.current?.clientHeight ?? 0,
                  segmentWidth,
                  segmentHeight,
                  "data-dtrum-allow": dataDtrumAllow,
                  "data-dtrum-mask": dataDtrumMask,
                  children: /* @__PURE__ */ jsxs(ExternalProviders, { ...externalProvidersProps, children: [
                    (label || icon) && /* @__PURE__ */ jsxs(
                      Flex,
                      {
                        ref: labelRef,
                        alignItems: "center",
                        alignSelf: "center",
                        className: labelContainer,
                        children: [
                          icon,
                          label
                        ]
                      }
                    ),
                    (isToolbarEnabled || value) && /* @__PURE__ */ jsx(
                      MeterBarLayoutTopRightSlot,
                      {
                        value,
                        isToolbarEnabled,
                        ref: toolbarRef
                      }
                    ),
                    /* @__PURE__ */ jsx("div", { className: chartContainer, ref: plotAreaRef, children: chart }),
                    min && /* @__PURE__ */ jsx("div", { ref: minLabelRef, className: minContainer, children: min }),
                    max && /* @__PURE__ */ jsx("div", { ref: maxLabelRef, className: maxContainer, children: max }),
                    legend && /* @__PURE__ */ jsx("div", { ref: legendRef, className: legendContainer, children: legend }),
                    loading ? /* @__PURE__ */ jsx(
                      LoaderOverlay,
                      {
                        gridArea: "1/1/2/2",
                        consumerStyle: consumerLoadingStyle
                      }
                    ) : null
                  ] })
                }
              )
            }
          )
        }
      )
    }
  );
});
MeterBarLayoutRoot.displayName = "MeterBarLayoutRoot";
const MeterBarLayout = Object.assign(MeterBarLayoutRoot, {
  Label,
  Min,
  Max,
  Chart,
  Value,
  Icon,
  Providers,
  Legend
});
export {
  MeterBarLayout,
  MeterBarLayoutRoot
};
