// @dynatrace/strato-components/tables/DataTable/features/ColumnVisibility/ColumnVisibility.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ColumnVisibility_exports = {};
__export(ColumnVisibility_exports, {
  DataTableColumnVisibility: () => DataTableColumnVisibility,
  getCanHide: () => getCanHide,
  useColumnVisibility: () => useColumnVisibility
});
module.exports = __toCommonJS(ColumnVisibility_exports);
var import_react = require("react");
var import_useControllableState = require("../../../../core/hooks/useControllableState.js");
var import_is_node_environment = require("../../../../core/utils/is-node-environment.js");
var import_useDataTableConfig = require("../../../hooks/useDataTableConfig.js");
var import_datatable_typeguards = require("../../hooks/useTable/datatable-typeguards.js");
var import_utils = require("../../hooks/useTable/utils.js");
var import_src = require("../../tanstack-table-core-override/src/index.js");
function getCanHide(table, column, isHideable) {
  if (table.options.enableHiding === false) {
    return false;
  }
  if (!column.columns || column.columns.length === 0) {
    return column.columnDef.enableHiding ?? true;
  }
  return column.columnDef.enableHiding === false ? false : isHideable(column.columns);
}
function isParentColumnHideable(childColumns) {
  return childColumns.every((c) => c.columnDef.enableHiding);
}
const DataTableColumnVisibility = {
  createTable(table) {
    table.getOrderedVisibleLeafColumns = (0, import_src.memo)(
      () => {
        const state = table.getState();
        return [
          state.columnVisibility,
          state.columnPinning,
          state.columnOrder,
          table.options.columns
        ];
      },
      () => {
        return [
          ...table.getLeftVisibleLeafColumns(),
          ...table.getCenterVisibleLeafColumns(),
          ...table.getRightVisibleLeafColumns()
        ];
      },
      { key: (0, import_is_node_environment.isDev)() && "getOrderedVisibleLeafColumns" }
    );
    table.getVisibleLeafColumnsWithoutBuiltin = (0, import_src.memo)(
      () => {
        const state = table.getState();
        return [
          state.columnVisibility,
          state.columnOrder,
          table.options.columns
        ];
      },
      () => {
        return table.getVisibleLeafColumns().filter((leafColumn) => !leafColumn.columnDef.meta?.builtin);
      },
      {
        key: (0, import_is_node_environment.isDev)() && "getVisibleLeafColumnsWithoutBuiltin"
      }
    );
    table.getCenterVisibleLeafColumnsWithoutBuiltIn = (0, import_src.memo)(
      () => {
        const state = table.getState();
        return [
          state.columnVisibility,
          state.columnPinning,
          state.columnOrder,
          table.options.columns
        ];
      },
      () => {
        return table.getCenterVisibleLeafColumns().filter((leafColumn) => !leafColumn.columnDef.meta?.builtin);
      },
      { key: (0, import_is_node_environment.isDev)() && "getCenterVisibleLeafColumnsWithoutBuildIn" }
    );
    table.getVisibleVirtualColumnIdsWithoutBuiltin = () => {
      const visibleLeafColumns = table.getVisibleLeafColumnsWithoutBuiltin();
      const virtualColumns = table.getColumnVirtualizer()?.getVirtualItems() ?? [];
      return virtualColumns.flatMap(
        (virtualItem) => visibleLeafColumns.at(virtualItem.index)?.id ?? []
      );
    };
  },
  createColumn(column, table) {
    column.getIsLastVisibleColumn = (0, import_src.memo)(
      () => {
        const state = table.getState();
        return [
          state.columnVisibility,
          state.columnPinning,
          table.options.columns
        ];
      },
      () => {
        const lastVisibleLeafColumnId = table.getRightVisibleLeafColumns().at(-1)?.id ?? table.getCenterVisibleLeafColumns().at(-1)?.id;
        return column.id === lastVisibleLeafColumnId || column.columns.some(
          (childColumn) => childColumn.id === lastVisibleLeafColumnId
        );
      },
      (0, import_utils.getMemoOptions)(
        table.options,
        "debugColumns",
        "column.getIsLastVisibleColumn"
      )
    );
    column.getVisibleLeafColumns = (0, import_src.memo)(
      () => {
        const state = table.getState();
        return [
          state.columnVisibility,
          state.columnOrder,
          table.options.columns
        ];
      },
      () => {
        return column.getLeafColumns().filter((leafColumn) => leafColumn.getIsVisible());
      },
      (0, import_utils.getMemoOptions)(
        table.options,
        "debugColumns",
        "column.getVisibleLeafColumns"
      )
    );
    column.getCenterVisibleLeafColumns = (0, import_src.memo)(
      () => {
        const state = table.getState();
        return [
          state.columnVisibility,
          state.columnPinning,
          state.columnOrder,
          table.options.columns
        ];
      },
      () => {
        return column.getVisibleLeafColumns().filter((leafColumn) => !leafColumn.getIsPinned());
      },
      (0, import_utils.getMemoOptions)(
        table.options,
        "debugColumns",
        "column.getCenterVisibleLeafColumns"
      )
    );
    column.getLeftVisibleLeafColumns = (0, import_src.memo)(
      () => {
        const state = table.getState();
        return [
          state.columnVisibility,
          state.columnPinning.left,
          state.columnOrder,
          table.options.columns
        ];
      },
      () => {
        const ids = new Set(column.getVisibleLeafColumns().map((c) => c.id));
        return table.getLeftVisibleLeafColumns().filter((c) => ids.has(c.id));
      },
      (0, import_utils.getMemoOptions)(
        table.options,
        "debugColumns",
        "column.getLeftVisibleLeafColumns"
      )
    );
    column.getRightVisibleLeafColumns = (0, import_src.memo)(
      () => {
        const state = table.getState();
        return [
          state.columnVisibility,
          state.columnPinning.right,
          state.columnOrder,
          table.options.columns
        ];
      },
      () => {
        const ids = new Set(column.getVisibleLeafColumns().map((c) => c.id));
        return table.getRightVisibleLeafColumns().filter((c) => ids.has(c.id));
      },
      (0, import_utils.getMemoOptions)(
        table.options,
        "debugColumns",
        "column.getRightVisibleLeafColumns"
      )
    );
    column.getCanHide = () => getCanHide(table, column, isParentColumnHideable);
  }
};
function useColumnVisibility(props, options) {
  const columnVisibilityConfig = (0, import_useDataTableConfig.useDataTableConfigElement)("columnVisibility");
  const [columnVisibility, onColumnVisibilityChange] = (0, import_useControllableState.useControllableState)({
    value: props.columnVisibility ?? columnVisibilityConfig,
    defaultValue: props.defaultColumnVisibility,
    onChange: props.onColumnVisibilityChange
  });
  options.state = { ...options.state, columnVisibility };
  options.onColumnVisibilityChange = onColumnVisibilityChange;
  options.originalColumnVisibility = (0, import_react.useMemo)(() => {
    return props.defaultColumnVisibility ?? getColumnVisibilityFromOriginalColumns(options.columns);
  }, [props.defaultColumnVisibility, options.columns]);
}
function getColumnVisibilityFromOriginalColumns(columnDefs) {
  const visibilityState = {};
  for (const columnDef of columnDefs) {
    if ((0, import_datatable_typeguards.isGroupColumnDef)(columnDef) && columnDef.columns) {
      const subColumnDefs = columnDef.columns;
      for (const subColumnDef of subColumnDefs) {
        if (subColumnDef.id) {
          visibilityState[subColumnDef.id] = true;
        }
      }
    } else if (columnDef.id) {
      visibilityState[columnDef.id] = true;
    }
  }
  return visibilityState;
}
