// @dynatrace/strato-components/esm/charts/timeseries/components/displays/area/accessors.js
import { safeGet } from "../../../../core/utils/safe-get.js";
import { getDatapointTime } from "../../../utils/timeseries.js";
import { getValueWithBoundaries } from "../shared/utils/clamp-value.js";
const definedAccessor = (dataPoint, datapointsMap) => !safeGet(datapointsMap, dataPoint).isGap;
const xAxisAccessor = (xScale, dataPointsMap, clampedScale) => (dataPoint) => {
  const { start, end, center } = safeGet(dataPointsMap, dataPoint);
  const timestamp = getDatapointTime(start, end, center);
  return getValueWithBoundaries(timestamp, xScale, clampedScale);
};
const y0AxisAccessor = (yScale, dataPointsMap) => (dataPoint) => {
  const dataPointMetadata = safeGet(dataPointsMap, dataPoint);
  const value = (dataPointMetadata.cumulativeValue ?? 0) - (dataPointMetadata.value ?? 0);
  return yScale(value);
};
const y1AxisAccessor = (yScale, dataPointsMap) => (dataPoint) => {
  const value = safeGet(dataPointsMap, dataPoint).cumulativeValue ?? 0;
  return yScale(value);
};
function y1(yScaleByUnit, seriesMetadata, datapointsMap) {
  return (dataPoint) => y1AxisAccessor(
    yScaleByUnit(seriesMetadata.unit),
    datapointsMap
  )(dataPoint) ?? 0;
}
function y0(yScaleByUnit, seriesMetadata, datapointsMap) {
  return (dataPoint) => y0AxisAccessor(
    yScaleByUnit(seriesMetadata.unit),
    datapointsMap
  )(dataPoint) ?? 0;
}
export {
  definedAccessor,
  xAxisAccessor,
  y0,
  y0AxisAccessor,
  y1,
  y1AxisAccessor
};
