// @dynatrace/strato-components/esm/charts/timeseries/components/displays/area/AreaChartDisplay.js
import { jsx } from "react/jsx-runtime";
import { useMemo } from "react";
import { definedAccessor, xAxisAccessor as x, y0, y1 } from "./accessors.js";
import { AreaCanvas } from "../../../../core/components/canvas/components/AreaCanvas.js";
import { shouldDisplayPoints } from "../../../../core/components/canvas/utils/display-points.js";
import { useGridPlotSize } from "../../../../core/components/chart-grid/hooks/useGridPlotSize.js";
import { DEFAULT_FILL_OPACITY } from "../../../../core/constants.js";
import { isIsolatedPoint } from "../../../../core/utils/is-isolated-point.js";
import { safeGet } from "../../../../core/utils/safe-get.js";
import { AREA_CHART_DISPLAY_ARIA_LABEL } from "../../../constants.js";
import { useSeriesCurve } from "../../../hooks/useSeriesCurve.js";
import { useTimeseriesFilteredData } from "../../../hooks/useTimeseriesFilteredData.js";
import { useYScales } from "../../../hooks/useYScales.js";
import { getYScaleByUnit } from "../../../utils/yAxis.js";
import { getClampedTimeScale } from "../shared/utils/clamp-value.js";
const AreaChartDisplay = ({
  series: timeseries,
  xScale
}) => {
  const yScale = useYScales();
  const curve = useSeriesCurve();
  const { width, height } = useGridPlotSize();
  const { seriesMap, datapointsMap } = useTimeseriesFilteredData();
  const areaSeries = useMemo(() => [...timeseries].reverse(), [timeseries]);
  const yScaleByUnit = (unit) => getYScaleByUnit(unit, yScale);
  const clampedScale = getClampedTimeScale(xScale);
  const isolatedAccessor = (displayPoints) => (datapoints, index) => displayPoints || isIsolatedPoint(
    datapoints,
    index,
    (point) => definedAccessor(point, datapointsMap)
  );
  return /* @__PURE__ */ jsx("figure", { "aria-label": AREA_CHART_DISPLAY_ARIA_LABEL, children: areaSeries.map((series) => {
    const seriesMetadata = safeGet(seriesMap, series);
    const displayPoints = shouldDisplayPoints(
      seriesMetadata.pointsDisplay,
      width,
      series.datapoints.length
    );
    return /* @__PURE__ */ jsx(
      AreaCanvas,
      {
        data: series.datapoints,
        canvasWidth: width,
        canvasHeight: height,
        x: x(xScale, datapointsMap, clampedScale),
        y0: y0(yScaleByUnit, seriesMetadata, datapointsMap),
        y1: y1(yScaleByUnit, seriesMetadata, datapointsMap),
        defined: (point) => definedAccessor(point, datapointsMap),
        displayScatterPlot: isolatedAccessor(displayPoints),
        color: () => seriesMetadata.color.Default,
        opacity: DEFAULT_FILL_OPACITY,
        curve,
        enableDotOverflow: true
      },
      seriesMetadata.id
    );
  }) });
};
AreaChartDisplay["displayName"] = "AreaChartDisplay";
export {
  AreaChartDisplay
};
