// @dynatrace/strato-components/core/time/timeframeStore/timeframeStoreProvider.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var timeframeStoreProvider_exports = {};
__export(timeframeStoreProvider_exports, {
  CalendarStoreCtx: () => CalendarStoreCtx,
  CalendarStoreProvider: () => CalendarStoreProvider,
  DateTimePickerStoreCtx: () => DateTimePickerStoreCtx,
  DateTimePickerStoreProvider: () => DateTimePickerStoreProvider,
  DateTimeRangePickerStoreCtx: () => DateTimeRangePickerStoreCtx,
  DateTimeRangePickerStoreProvider: () => DateTimeRangePickerStoreProvider,
  TimeframeSelectorStoreCtx: () => TimeframeSelectorStoreCtx,
  TimeframeSelectorStoreProvider: () => TimeframeSelectorStoreProvider
});
module.exports = __toCommonJS(timeframeStoreProvider_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
const CalendarStoreCtx = (0, import_react.createContext)(null);
const DateTimePickerStoreCtx = (0, import_react.createContext)(null);
const DateTimeRangePickerStoreCtx = (0, import_react.createContext)(null);
const TimeframeSelectorStoreCtx = (0, import_react.createContext)(null);
function CalendarStoreProvider({
  store,
  children
}) {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(CalendarStoreCtx.Provider, { value: store, children });
}
function DateTimePickerStoreProvider({
  store,
  children
}) {
  const calStore = store;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(CalendarStoreCtx.Provider, { value: calStore, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(DateTimePickerStoreCtx.Provider, { value: store, children }) });
}
function DateTimeRangePickerStoreProvider({
  store,
  children
}) {
  const calStore = store;
  const dtpStore = store;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(CalendarStoreCtx.Provider, { value: calStore, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(DateTimePickerStoreCtx.Provider, { value: dtpStore, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(DateTimeRangePickerStoreCtx.Provider, { value: store, children }) }) });
}
function TimeframeSelectorStoreProvider({
  store,
  children
}) {
  const calStore = store;
  const dtpStore = store;
  const dtrpStore = store;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(CalendarStoreCtx.Provider, { value: calStore, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(DateTimePickerStoreCtx.Provider, { value: dtpStore, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(DateTimeRangePickerStoreCtx.Provider, { value: dtrpStore, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(TimeframeSelectorStoreCtx.Provider, { value: store, children }) }) }) });
}
