// @dynatrace/strato-components/charts/core/components/annotations/components/canvas/utils/calculate-marker-width-and-content.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var calculate_marker_width_and_content_exports = {};
__export(calculate_marker_width_and_content_exports, {
  calculateMarkerWidthAndContent: () => calculateMarkerWidthAndContent
});
module.exports = __toCommonJS(calculate_marker_width_and_content_exports);
var import_check_marker_fit_centered = require("./check-marker-fit-centered.js");
var import_get_default_value_marker_pointer_position = require("./get-default-value-marker-pointer-position.js");
var import_utils = require("../../../../../../../typography/utils.js");
var import_chars_font_width_map = require("../../../../text-ellipsis/chars-font-width-map.js");
var import_truncate_string = require("../../../../text-ellipsis/truncate-string.js");
var import_constants = require("../../../constants.js");
const MIN_TEXT_TO_RENDER = `W${import_utils.ELLIPSIS_CHARACTER}W`;
const RANGE_MARKER_WITH_SYMBOL_MIN_WIDTH = 20;
function getAvailableWidth(width) {
  return width - import_constants.MARKER_CONTENT_TO_BORDER_PADDING * 2;
}
function getSanitizedContent(content, availableWidth, truncateMode) {
  const minimumContentAllowedWidth = (0, import_truncate_string.calculateWidth)(
    MIN_TEXT_TO_RENDER,
    import_chars_font_width_map.charsFontWidth12pixelsSize500Map
  );
  if (minimumContentAllowedWidth > availableWidth) {
    return "";
  }
  return (0, import_truncate_string.truncateString)(
    content,
    availableWidth,
    import_chars_font_width_map.charsFontWidth12pixelsSize500Map,
    truncateMode
  );
}
function resolveConstrainedLayout(content, xPosition, markerWidth, trackLeftBoundary, trackRightBoundary, isRange, truncateMode) {
  const spaceToTheRight = trackRightBoundary - xPosition;
  const moreSpaceRight = spaceToTheRight > xPosition;
  const constrainedWidth = moreSpaceRight ? Math.min(spaceToTheRight, markerWidth) : xPosition - trackLeftBoundary;
  return {
    sanitizedContent: getSanitizedContent(
      content,
      getAvailableWidth(constrainedWidth),
      truncateMode
    ),
    calculatedMarkerWidth: constrainedWidth,
    valueMarkerOrientation: moreSpaceRight ? "left" : "right",
    calculatedX: !moreSpaceRight && isRange ? trackLeftBoundary : xPosition
  };
}
function resolveExpandedStringLayout(content, width, isRange, textOverflow, boundaries, truncateMode) {
  const {
    xPosition,
    trackLeftBoundary,
    trackRightBoundary,
    markerWidth,
    defaultPointer
  } = boundaries;
  if (isRange && textOverflow !== "expand") {
    return {
      sanitizedContent: getSanitizedContent(
        content,
        getAvailableWidth(width),
        truncateMode
      ),
      calculatedMarkerWidth: width,
      valueMarkerOrientation: defaultPointer,
      calculatedX: xPosition
    };
  }
  const canExpandCenter = (0, import_check_marker_fit_centered.checkMarkerFitCentered)(
    xPosition,
    markerWidth,
    trackLeftBoundary,
    trackRightBoundary
  );
  if (canExpandCenter) {
    const adjustedX = isRange ? width / 2 + xPosition - markerWidth / 2 : xPosition;
    return {
      sanitizedContent: content,
      calculatedMarkerWidth: markerWidth,
      valueMarkerOrientation: defaultPointer,
      calculatedX: adjustedX
    };
  }
  if (xPosition - markerWidth >= trackLeftBoundary) {
    const adjustedX = !isRange && textOverflow === "expand" ? xPosition - markerWidth : xPosition;
    return {
      sanitizedContent: content,
      calculatedMarkerWidth: markerWidth,
      valueMarkerOrientation: "right",
      calculatedX: adjustedX
    };
  }
  if (xPosition + markerWidth <= trackRightBoundary) {
    return {
      sanitizedContent: content,
      calculatedMarkerWidth: markerWidth,
      valueMarkerOrientation: "left",
      calculatedX: xPosition
    };
  }
  return resolveConstrainedLayout(
    content,
    xPosition,
    markerWidth,
    trackLeftBoundary,
    trackRightBoundary,
    isRange,
    truncateMode
  );
}
function calculateMarkerWidthAndContent({
  trackWidth,
  trackLeftBoundary,
  content,
  width,
  x,
  isRange,
  textOverflow,
  truncateMode
}) {
  const trackRightBoundary = trackWidth + trackLeftBoundary;
  const xPosition = trackLeftBoundary + x;
  const defaultPointer = (0, import_get_default_value_marker_pointer_position.getDefaultValueMarkerPointerPosition)(
    xPosition,
    trackLeftBoundary,
    trackRightBoundary
  );
  if (typeof content !== "string") {
    const sanitizedContent = isRange && RANGE_MARKER_WITH_SYMBOL_MIN_WIDTH > getAvailableWidth(width) ? "" : content;
    return {
      sanitizedContent,
      calculatedMarkerWidth: width,
      valueMarkerOrientation: defaultPointer,
      calculatedX: xPosition
    };
  }
  const textWidth = (0, import_truncate_string.calculateWidth)(content, import_chars_font_width_map.charsFontWidth12pixelsSize500Map);
  const markerWidth = Math.max(
    textWidth + import_constants.MARKER_CONTENT_TO_BORDER_PADDING * 2,
    width
  );
  if (markerWidth <= width) {
    return {
      sanitizedContent: content,
      calculatedMarkerWidth: width,
      valueMarkerOrientation: defaultPointer,
      calculatedX: xPosition
    };
  }
  const boundaries = {
    xPosition,
    trackLeftBoundary,
    trackRightBoundary,
    markerWidth,
    defaultPointer
  };
  return resolveExpandedStringLayout(
    content,
    width,
    isRange,
    textOverflow,
    boundaries,
    truncateMode
  );
}
