// @dynatrace/strato-components/charts/core/components/annotations/components/canvas/utils/get-marker-brightness-percentage-based-on-state.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var get_marker_brightness_percentage_based_on_state_exports = {};
__export(get_marker_brightness_percentage_based_on_state_exports, {
  getMarkerBrightnessPercentageBasedOnState: () => getMarkerBrightnessPercentageBasedOnState
});
module.exports = __toCommonJS(get_marker_brightness_percentage_based_on_state_exports);
const MARKER_DEFAULT_BRIGHTNESS_PERCENTAGE = "100%";
const mapStateBrightnessPercentage = {
  dark: {
    default: MARKER_DEFAULT_BRIGHTNESS_PERCENTAGE,
    hovered: "110%",
    active: "120%",
    inactive: "60%"
  },
  light: {
    default: MARKER_DEFAULT_BRIGHTNESS_PERCENTAGE,
    hovered: "90%",
    active: "80%",
    inactive: "140%"
  }
};
const getMarkerBrightnessPercentageBasedOnState = (theme, state) => mapStateBrightnessPercentage[theme][state];
