// @dynatrace/strato-components/esm/charts/core/components/copy-to-clipboard/copy-to-clipboard-messages.js
import { defineMessages } from "react-intl";
const COPY_TO_CLIPBOARD_MESSAGES = defineMessages({
  copyActionName: {
    id: "6vA3zfYytiSPVBZV",
    defaultMessage: "Copy name",
    description: "Text in a tooltip shown when clicking a tooltip item and hovering the copy button next to its name"
  },
  copyCoordinates: {
    id: "Rrlag1wRQFMpYHX7",
    defaultMessage: "Copy coordinates",
    description: "Text in a tooltip shown when clicking a geo map tooltip item and hovering the copy button next to its coordinates"
  },
  tooltipActionCopy: {
    id: "FxsZTOljj8LWcZO0",
    defaultMessage: "Copy",
    description: "Text in a tooltip shown by default when hovering"
  },
  actionCopied: {
    id: "aYOFXw5HKPdKKWSu",
    defaultMessage: "Copied",
    description: "Text in a tooltip shown when clicking a tooltip series and copying its series name"
  },
  copyField: {
    id: "XOE5zKTRmqYLBnRf",
    defaultMessage: "Copy {field}",
    description: "Label for the copy action in the tooltip, where field is the name of the field to be copied"
  }
});
export {
  COPY_TO_CLIPBOARD_MESSAGES
};
