// @dynatrace/strato-components/esm/forms/forms-core/components/form-field/FormFieldContextProvider.js
import { jsx } from "react/jsx-runtime";
import { useReducer } from "react";
import { useId } from "../../../../core/hooks/useId.js";
import { parseBoolean as _parseBoolean } from "../../../../core/utils/parse-boolean.js";
import { _FormFieldContext as FormFieldContext } from "../../contexts/FormFieldContext.js";
import { useFormFieldContext } from "../../hooks/useFormFieldContext.js";
import { _formFieldReducer } from "../../state/formfield-state-reducer.js";
const FormFieldContextProvider = (props) => {
  const generatedId = useId();
  const {
    children,
    controlId: propId = generatedId,
    disabled: propDisabled,
    required: propRequired,
    ariaDisabled: ariaDisabledProp,
    readOnly: propReadOnly
  } = props;
  const {
    disabled: contextDisabled,
    ariaDisabled: contextAriaDisabled,
    readOnly: contextReadOnly
  } = useFormFieldContext();
  const [formFieldState, formFieldDispatch] = useReducer(_formFieldReducer, {});
  return /* @__PURE__ */ jsx(
    FormFieldContext.Provider,
    {
      value: {
        controlId: formFieldState.controlId || propId,
        disabled: formFieldState.disabled ?? propDisabled ?? contextDisabled,
        required: formFieldState.required ?? propRequired,
        readOnly: formFieldState.readOnly ?? propReadOnly ?? contextReadOnly,
        dispatch: formFieldDispatch,
        ariaDisabled: formFieldState.ariaDisabled ?? _parseBoolean(ariaDisabledProp) ?? contextAriaDisabled
      },
      children
    }
  );
};
FormFieldContextProvider.displayName = "FormFieldContextProvider";
export {
  FormFieldContextProvider
};
