// @dynatrace/strato-components/esm/forms/forms-core/components/form-field/FormField.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import { forwardRef } from "react";
import * as styles from "./FormField.sty.js";
import { FormFieldContextProvider } from "./FormFieldContextProvider.js";
import { FormFieldMessageContextProvider } from "./FormFieldMessageContextProvider.js";
import { useBehavioralTrackingProps } from "../../../../core/hooks/useBehavioralTrackingProps.js";
import { useId } from "../../../../core/hooks/useId.js";
import { mergeProps } from "../../../../core/utils/merge-props.js";
import { parseBoolean as _parseBoolean } from "../../../../core/utils/parse-boolean.js";
import { Flex } from "../../../../layouts/flex/Flex.js";
const COMPONENT_NAME = "FormField";
const FormField = /* @__PURE__ */ forwardRef((formFieldProps, forwardedRef) => {
  const generatedId = useId();
  const {
    children,
    id: propId = generatedId,
    disabled,
    required,
    readOnly,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "aria-disabled": ariaDisabledProp,
    ...remainingProps
  } = formFieldProps;
  const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
  return /* @__PURE__ */ jsx(
    FormFieldContextProvider,
    {
      controlId: propId,
      disabled,
      ariaDisabled: _parseBoolean(ariaDisabledProp),
      required,
      readOnly,
      children: /* @__PURE__ */ jsx(FormFieldMessageContextProvider, { children: /* @__PURE__ */ jsx(
        Flex,
        {
          ref: forwardedRef,
          "data-testid": dataTestId,
          "data-dtrum-mask": dataDtrumMask,
          "data-dtrum-allow": dataDtrumAllow,
          className: clsx(styles.formField, consumerClassName),
          gap: 4,
          flexDirection: "column",
          style: consumerStyle,
          ...mergeProps(restProps, behavioralTrackingProps),
          "data-dt-component": COMPONENT_NAME,
          children
        }
      ) })
    }
  );
});
FormField.displayName = COMPONENT_NAME;
export {
  FormField
};
