// @dynatrace/strato-components/esm/forms/forms-core/components/form-field/FormFieldMessageContextProvider.js
import { jsx } from "react/jsx-runtime";
import {
  forwardRef,
  useCallback,
  useReducer,
  useMemo
} from "react";
import { useId } from "../../../../core/hooks/useId.js";
import { FormFieldMessageContext } from "../../contexts/FormFieldMessageContext.js";
import { useFormFieldContext } from "../../hooks/useFormFieldContext.js";
import { useFormFieldMessagesContext } from "../../hooks/useFormFieldMessagesContext.js";
import {
  formFieldMessagesReducer
} from "../../state/formfield-messages-state-reducer.js";
const FormFieldMessageContextProvider = (formFieldProps) => {
  const generatedId = useId();
  const {
    children,
    id: propId = generatedId,
    disabled: propDisabled
  } = formFieldProps;
  const formFieldState = useFormFieldContext();
  const formFieldMessageReducerCB = useCallback(
    (state2, action) => formFieldMessagesReducer(state2, action),
    []
  );
  const [state, dispatch] = useReducer(formFieldMessageReducerCB, {
    customMessages: [],
    formControlMessages: [],
    showError: false,
    isTouched: false,
    isValid: void 0,
    controlIds: [formFieldState.controlId ?? propId]
  });
  const formFieldMessageContextValue = useMemo(
    () => ({
      disabled: formFieldState.disabled || propDisabled,
      controlId: formFieldState.controlId || propId,
      dispatch,
      state
    }),
    [
      formFieldState.disabled,
      formFieldState.controlId,
      propDisabled,
      propId,
      dispatch,
      state
    ]
  );
  return /* @__PURE__ */ jsx(FormFieldMessageContext.Provider, { value: formFieldMessageContextValue, children });
};
function withFormFieldMessageContextProviderWithGenerics(OriginalComponent) {
  const WrappedComponent = forwardRef((props, forwardedRef) => {
    const { dispatch } = useFormFieldMessagesContext();
    if (dispatch !== void 0) {
      return /* @__PURE__ */ jsx(OriginalComponent, { ref: forwardedRef, ...props });
    }
    const { disabled, id } = props;
    return /* @__PURE__ */ jsx(FormFieldMessageContextProvider, { disabled, id, children: /* @__PURE__ */ jsx(OriginalComponent, { ref: forwardedRef, ...props }) });
  });
  return WrappedComponent;
}
function withFormFieldMessageContextProvider(OriginalComponent) {
  const WithFormFieldMessageContextProvider = forwardRef(
    (props, forwardedRef) => {
      const { dispatch } = useFormFieldMessagesContext();
      if (dispatch !== void 0) {
        return /* @__PURE__ */ jsx(OriginalComponent, { ref: forwardedRef, ...props });
      }
      const { disabled, id } = props;
      return /* @__PURE__ */ jsx(FormFieldMessageContextProvider, { disabled, id, children: /* @__PURE__ */ jsx(OriginalComponent, { ref: forwardedRef, ...props }) });
    }
  );
  return WithFormFieldMessageContextProvider;
}
FormFieldMessageContextProvider.displayName = "FormFieldMessageContextProvider";
export {
  FormFieldMessageContextProvider,
  withFormFieldMessageContextProvider,
  withFormFieldMessageContextProviderWithGenerics
};
