// @dynatrace/strato-components/tables/DataTable/features/ColumnOrder/ColumnOrder.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ColumnOrder_exports = {};
__export(ColumnOrder_exports, {
  DataTableColumnOrder: () => DataTableColumnOrder,
  useColumnOrder: () => useColumnOrder
});
module.exports = __toCommonJS(ColumnOrder_exports);
var import_react_table = require("@tanstack/react-table");
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_column_order_utils = require("./column-order-utils.js");
var import_column_order_validation = require("./column-order-validation.js");
var import_useControllableState = require("../../../../core/hooks/useControllableState.js");
var import_utils = require("../../hooks/useTable/utils.js");
const DataTableColumnOrder = {
  createTable(table) {
    table._getOrderColumnsFn = (0, import_react_table.memo)(
      () => {
        const tableState = table.getState();
        return [
          tableState.columnOrder,
          tableState.grouping,
          table.options.groupedColumnMode
        ];
      },
      (columnOrder, grouping, groupedColumnMode) => (columns) => {
        const prefixColumns = [];
        const suffixColumns = [];
        let orderedColumns = columns;
        if (columnOrder.length) {
          const columnsMap = /* @__PURE__ */ new Map();
          columns.forEach((column) => {
            const builtinPlacement = column.columnDef.meta?.builtinPlacement;
            if (builtinPlacement === "prefix") {
              prefixColumns.push(column);
            } else if (builtinPlacement === "suffix") {
              suffixColumns.push(column);
            } else {
              columnsMap.set(column.id, column);
            }
          });
          orderedColumns = [
            ...prefixColumns,
            ...columnOrder.flatMap((id) => columnsMap.get(id) ?? []),
            ...suffixColumns
          ];
        }
        return (0, import_react_table.orderColumns)(
          orderedColumns,
          grouping,
          groupedColumnMode
        );
      },
      (0, import_utils.getMemoOptions)(table.options, "debugCells", "cell._getOrderColumnsFn")
    );
    table.getLeafColumnsWithoutBuiltin = () => {
      return table.getAllLeafColumns().filter((leafColumn) => !leafColumn.columnDef.meta?.builtin);
    };
    table.setDraggedHeaderColumn = (column) => {
      table.setState((state) => ({
        ...state,
        draggedHeaderColumn: column
      }));
    };
  },
  createHeader(header, table) {
    header.getColumnDragAndDropMode = () => {
      const isDragAndDropDisabled = table.options.enableColumnOrdering !== true || header.column.columnDef.enableUserActions === false || header.getIsPinned() || !header.column.getCanMoveColumnLeft() && !header.column.getCanMoveColumnRight();
      if (isDragAndDropDisabled) {
        return "none";
      }
      if (!header.column.getCanColumnActions() && !header.getCanSortByHeaderClick()) {
        return "only";
      }
      return "enabled";
    };
    header.getIsDragAndDropDisabled = () => {
      const { draggedHeaderColumn } = table.getState();
      const draggedHeaderParentId = draggedHeaderColumn?.parent?.id;
      const draggedHeaderId = draggedHeaderColumn?.id;
      const isPinnedHeader = !!header.getIsPinned();
      if (draggedHeaderId && isPinnedHeader) {
        return true;
      }
      if (draggedHeaderParentId === void 0 || draggedHeaderParentId === header.column.id || draggedHeaderParentId === header.column.parent?.id) {
        return false;
      } else {
        return true;
      }
    };
  },
  createColumn(column, table) {
    column.getCanMoveColumnLeft = () => {
      const columnsToCompare = (0, import_column_order_utils.getColumnsToCompareForMove)(table, column);
      return (0, import_column_order_utils.canMoveColumnLeft)(column, columnsToCompare);
    };
    column.getCanMoveColumnRight = () => {
      const columnsToCompare = (0, import_column_order_utils.getColumnsToCompareForMove)(table, column);
      return (0, import_column_order_utils.canMoveColumnRight)(column, columnsToCompare);
    };
    column.moveColumnLeft = () => {
      const leafColumns = table.getLeafColumnsWithoutBuiltin();
      const currentColumnOrder = table.getState().columnOrder;
      const columnsPinned = table.getState().columnPinning;
      const newColumnOrder = (0, import_column_order_utils.moveColumnLeft)(
        column,
        currentColumnOrder,
        columnsPinned,
        leafColumns
      );
      table.setColumnOrder(newColumnOrder);
    };
    column.moveColumnRight = () => {
      const leafColumns = table.getLeafColumnsWithoutBuiltin();
      const currentColumnOrder = table.getState().columnOrder;
      const columnsPinned = table.getState().columnPinning;
      const newColumnOrder = (0, import_column_order_utils.moveColumnRight)(
        column,
        currentColumnOrder,
        columnsPinned,
        leafColumns
      );
      table.setColumnOrder(newColumnOrder);
    };
  }
};
function useColumnOrder(props, options) {
  options.enableColumnOrdering = false;
  const { columnOrder: columnOrderFromDefinition, columnGroups } = (0, import_column_order_validation.extractColumnOrderAndGroups)(props.columns);
  const lastValidColumnOrder = (0, import_react.useRef)(
    columnOrderFromDefinition
  );
  const lastColumnOrderFromDefinition = (0, import_react.useRef)(
    columnOrderFromDefinition
  );
  const [columnOrder, setColumnOrder] = (0, import_useControllableState.useControllableState)({
    value: props.columnOrder,
    // if no order was specified, initialize with column definition order
    defaultValue: props.defaultColumnOrder || columnOrderFromDefinition
  });
  const onColumnOrderChangeCallback = (0, import_react.useCallback)(
    (updater, trigger = "user") => {
      const nextState = (0, import_react_table.functionalUpdate)(
        updater,
        options.state?.columnOrder ?? columnOrder
      );
      if (!(0, import_lodash_es.isEqual)(nextState, columnOrder)) {
        setColumnOrder(nextState);
        props.onColumnOrderChange?.(nextState, trigger);
      }
    },
    [columnOrder, options.state?.columnOrder, props, setColumnOrder]
  );
  if (!props.columnOrdering) {
    return;
  }
  if (props.columnOrder === void 0 && !(0, import_lodash_es.isEqual)(lastColumnOrderFromDefinition.current, columnOrderFromDefinition)) {
    onColumnOrderChangeCallback(columnOrderFromDefinition, "internal");
    lastColumnOrderFromDefinition.current = columnOrderFromDefinition;
  } else if (!(0, import_lodash_es.isEqual)(columnOrder, lastValidColumnOrder.current)) {
    const validColumnOrder = (0, import_column_order_validation.validateColumnOrder)(
      columnOrder,
      columnOrderFromDefinition,
      columnGroups
    );
    if (validColumnOrder !== null) {
      onColumnOrderChangeCallback(validColumnOrder, "internal");
      lastValidColumnOrder.current = validColumnOrder;
    }
  }
  options.enableColumnOrdering = true;
  options.state = { ...options.state, columnOrder };
  options.onColumnOrderChange = onColumnOrderChangeCallback;
  options.originalColumnOrder = props.defaultColumnOrder ?? lastValidColumnOrder.current;
}
