// @dynatrace/strato-components/tables/DataTable/features/ColumnOrder/ColumnOrderDragAndDropContext.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ColumnOrderDragAndDropContext_exports = {};
__export(ColumnOrderDragAndDropContext_exports, {
  ColumnOrderDragAndDropContext: () => ColumnOrderDragAndDropContext
});
module.exports = __toCommonJS(ColumnOrderDragAndDropContext_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_core = require("@dnd-kit/core");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_dom = require("react-dom");
var import_column_order_dnd_utils = require("./column-order-dnd-utils.js");
var styles = __toESM(require("./ColumnOrder.sty.js"));
var import_data_table_border_width = require("../../../utils/data-table-border-width.js");
var import_DataTableContext = require("../../contexts/DataTableContext.js");
const DRAGGED_HEADER_DEFAULT_WIDTH = 100;
const DRAGGED_HEADER_DEFAULT_HEIGHT = 37;
function ColumnOrderDragAndDropContext({
  children
}) {
  const table = (0, import_DataTableContext.useDataTable)();
  const totalHeaderHeight = table.getTotalHeaderHeight();
  const visualVariantOptions = table.getVisualVariantOptions();
  const containedVariant = visualVariantOptions.contained;
  const hasSubHeaders = table.getHeaderGroups().length > 1;
  const verticalDividers = visualVariantOptions.verticalDividers || hasSubHeaders;
  const isColumnPinningEnabled = table.options.enableColumnPinning;
  const pinnedColumns = table.getState().columnPinning;
  const virtualizationContainer = table.getRowVirtualizationContainer();
  const [dragOverlayContent, setDragOverlayContent] = (0, import_react.useState)(null);
  const pinnedDraggableAreaBoundsRef = (0, import_react.useRef)({ left: void 0, right: void 0 });
  const [draggedHeaderSize, setDraggedHeaderSize] = (0, import_react.useState)({
    width: DRAGGED_HEADER_DEFAULT_WIDTH,
    height: DRAGGED_HEADER_DEFAULT_HEIGHT
  });
  const sensors = (0, import_core.useSensors)(
    (0, import_core.useSensor)(import_core.MouseSensor, {}),
    (0, import_core.useSensor)(import_core.TouchSensor, {})
  );
  const restrictColumnDragAndDrop = (0, import_react.useMemo)(
    () => (0, import_column_order_dnd_utils.createRestrictColumnDragAndDrop)(
      totalHeaderHeight,
      containedVariant,
      pinnedColumns,
      hasSubHeaders,
      virtualizationContainer
    ),
    [
      totalHeaderHeight,
      virtualizationContainer,
      containedVariant,
      pinnedColumns,
      hasSubHeaders
    ]
  );
  const placeholderElement = (0, import_react.useRef)(null);
  const scrollFrameRef = (0, import_react.useRef)(null);
  const pointerXRef = (0, import_react.useRef)(0);
  const trackPointerXRef = (0, import_react.useRef)(null);
  const parentHeaderElementRef = (0, import_react.useRef)(null);
  const stopScrollLoop = (0, import_react.useCallback)(() => {
    if (scrollFrameRef.current !== null) {
      cancelAnimationFrame(scrollFrameRef.current);
      scrollFrameRef.current = null;
    }
  }, []);
  const startScrollLoop = (0, import_react.useCallback)(() => {
    if (scrollFrameRef.current !== null) {
      return;
    }
    const loop = () => {
      if (!virtualizationContainer) {
        scrollFrameRef.current = null;
        return;
      }
      const mouseX = pointerXRef.current;
      const containerRect = virtualizationContainer.getBoundingClientRect();
      const rightBound = pinnedDraggableAreaBoundsRef.current.right ?? containerRect.right;
      const leftBound = pinnedDraggableAreaBoundsRef.current.left ?? containerRect.left;
      const rightPenetration = mouseX - (rightBound - import_column_order_dnd_utils.SCROLL_PROXIMITY_ZONE);
      const leftPenetration = leftBound + import_column_order_dnd_utils.SCROLL_PROXIMITY_ZONE - mouseX;
      const { scrollLeft, scrollWidth, clientWidth } = virtualizationContainer;
      let maxScrollRight = Infinity;
      let maxScrollLeft = Infinity;
      const parentEl = parentHeaderElementRef.current;
      if (parentEl) {
        const parentRect = parentEl.getBoundingClientRect();
        maxScrollRight = Math.max(0, parentRect.right - rightBound);
        maxScrollLeft = Math.max(0, leftBound - parentRect.left);
      }
      if (rightPenetration > 0 && leftPenetration <= 0) {
        const scrollSpeed = (0, import_column_order_dnd_utils.getCustomScrollSpeed)(rightPenetration);
        const canScrollRight = scrollLeft < scrollWidth - clientWidth;
        if (canScrollRight && maxScrollRight > 0) {
          virtualizationContainer.scrollLeft += Math.min(
            scrollSpeed,
            Math.ceil(maxScrollRight)
          );
        }
      } else if (leftPenetration > 0 && rightPenetration <= 0) {
        const scrollSpeed = (0, import_column_order_dnd_utils.getCustomScrollSpeed)(leftPenetration);
        const canScrollLeft = scrollLeft > 0;
        if (canScrollLeft && maxScrollLeft > 0) {
          virtualizationContainer.scrollLeft -= Math.min(
            scrollSpeed,
            Math.ceil(maxScrollLeft)
          );
        }
      }
      scrollFrameRef.current = requestAnimationFrame(loop);
    };
    scrollFrameRef.current = requestAnimationFrame(loop);
  }, [virtualizationContainer]);
  const cleanupDragState = (0, import_react.useCallback)(() => {
    stopScrollLoop();
    parentHeaderElementRef.current = null;
    if (trackPointerXRef.current) {
      window.removeEventListener("pointermove", trackPointerXRef.current);
      trackPointerXRef.current = null;
    }
  }, [stopScrollLoop]);
  const handleDragStart = ({ active }) => {
    const draggedColumn = active.data.current?.column;
    if (isColumnPinningEnabled) {
      pinnedDraggableAreaBoundsRef.current = (0, import_column_order_dnd_utils.getDraggableAreaBounds)(
        pinnedColumns,
        virtualizationContainer,
        hasSubHeaders ? 1 : 0
      );
      trackPointerXRef.current = (e) => {
        pointerXRef.current = e.clientX;
      };
      window.addEventListener("pointermove", trackPointerXRef.current, {
        passive: true
      });
      if (draggedColumn.parent) {
        const parentId = draggedColumn.parent.id;
        parentHeaderElementRef.current = virtualizationContainer?.querySelector(
          `[role="columnheader"][data-header-row="0"][data-columnid=${CSS.escape(
            parentId
          )}]${import_column_order_dnd_utils.unpinnedHeaderFilter}`
        ) ?? null;
      }
      startScrollLoop();
    }
    table.setDraggedHeaderColumn(draggedColumn);
    setDraggedHeaderSize(table.getHeaderSize(active.id));
    setDragOverlayContent(active.data.current?.portalRef);
    const { headerGroupElement, draggedHeaderElement } = (0, import_column_order_dnd_utils.getDraggedHeaderElement)(String(active.id), virtualizationContainer);
    if (headerGroupElement === null || !(headerGroupElement instanceof HTMLElement)) {
      return;
    }
    if (draggedHeaderElement === null || !(draggedHeaderElement instanceof HTMLElement)) {
      return;
    }
    const { gridArea } = draggedHeaderElement.style;
    const { width, height } = draggedHeaderElement.getBoundingClientRect();
    if (!gridArea) {
      return;
    }
    const verticalDividerOptions = verticalDividers && !draggedColumn.getIsFirstVisibleUnpinnedColumn() ? {
      isParentColumn: draggedHeaderElement.getAttribute("data-header-row") === "0",
      heightOffset: headerGroupElement.offsetHeight - draggedHeaderElement.offsetHeight
    } : void 0;
    placeholderElement.current = (0, import_column_order_dnd_utils.createHeaderPlaceholder)(
      gridArea,
      width,
      height,
      verticalDividerOptions
    );
    headerGroupElement.appendChild(placeholderElement.current);
  };
  const handleDragOver = (event) => {
    (0, import_column_order_dnd_utils.indicateInsertPosition)(event, virtualizationContainer);
  };
  const handleDragEnd = ({
    active,
    over
  }) => {
    setDragOverlayContent(null);
    cleanupDragState();
    (0, import_column_order_dnd_utils.updateColumnOrderOnDragEnd)({
      active,
      over,
      table,
      virtualizationContainer
    });
    requestAnimationFrame(() => {
      placeholderElement.current?.remove();
    });
  };
  (0, import_react.useEffect)(() => {
    return () => {
      cleanupDragState();
      placeholderElement.current?.remove();
    };
  }, [stopScrollLoop, cleanupDragState]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_core.DndContext,
    {
      sensors,
      collisionDetection: import_column_order_dnd_utils.customCollisionDetection,
      modifiers: [restrictColumnDragAndDrop],
      autoScroll: !isColumnPinningEnabled,
      accessibility: { container: table.getRowVirtualizationContainer() },
      onDragStart: handleDragStart,
      onDragOver: handleDragOver,
      onDragEnd: handleDragEnd,
      // We need to portal the drag and drop overlay to escape CSS transforms on ancestor elements.
      // This ensures transforms on parent containers don't affect the position of the drag overlay.
      children: [
        (0, import_react_dom.createPortal)(
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_core.DragOverlay,
            {
              "data-drag-overlay": true,
              dropAnimation: null,
              className: (0, import_clsx.default)(styles.columnDndOverlay),
              style: {
                // the observed header size includes a border which we don't want in the overlay
                height: draggedHeaderSize.height - (0, import_data_table_border_width.getDataTableBorderWidth)(),
                width: draggedHeaderSize.width
              },
              children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { ref: dragOverlayContent })
            }
          ),
          document.body
        ),
        children
      ]
    }
  );
}
