// @dynatrace/strato-components/esm/charts/histogram/utils/histogram-configuration.js
import { merge } from "lodash-es";
import { getToolbarConfiguration } from "./get-toolbar-configuration.js";
import {
  DEFAULT_INFINITE_ZOOM,
  DEFAULT_TRUNCATION_MODE,
  DEFAULT_VALUE_REPRESENTATION
} from "../../core/constants.js";
import { isValidColorRule } from "../../core/slots/ColorRule.js";
import {
  isCustomColorPalette
} from "../../core/types/color-palette.js";
import { isValidCSSColor } from "../../core/utils/getValidColor.js";
import { getInteractions } from "../../core/utils/interactions.js";
import { isValidColorPalette } from "../../core/utils/is-valid-color-palette.js";
import { isValidLegendPosition } from "../../core/utils/is-valid-legend-position.js";
import {
  DEFAULT_SCALE_FALLBACK_COLOR,
  defaultColorPalette,
  defaultHistogramChartConfig,
  defaultLegend,
  defaultTooltipVariant,
  defaultXAxis,
  defaultYAxis
} from "../constants/defaults.js";
import {
  histogramTooltipVariantOptions
} from "../types/histogram.js";
function getColorPaletteConfiguration(contextColorPalette, propsColorPalette) {
  const colorPalette = propsColorPalette ?? contextColorPalette ?? defaultColorPalette;
  return isCustomColorPalette(colorPalette) || isValidColorPalette(colorPalette) ? colorPalette : defaultHistogramChartConfig.colorPalette ?? defaultColorPalette;
}
function getLegendConfiguration(contextLegend, slotsLegend) {
  const legendConfig = merge(
    {},
    defaultLegend,
    contextLegend,
    slotsLegend
  );
  legendConfig.position = isValidLegendPosition(legendConfig.position) ? legendConfig.position : defaultLegend.position;
  return legendConfig;
}
const isValidHistogramTooltipVariant = (tooltipVariant) => histogramTooltipVariantOptions.includes(
  tooltipVariant
);
function getTooltipConfiguration(contextTooltip, slotsTooltip) {
  const tooltip = merge(
    {},
    { variant: defaultTooltipVariant },
    contextTooltip,
    slotsTooltip
  );
  tooltip.variant = isValidHistogramTooltipVariant(tooltip.variant) ? tooltip.variant : defaultTooltipVariant;
  return tooltip;
}
const sanitizeYAxisConfig = (yAxes) => {
  const reversedYAxes = yAxes.reverse();
  const leftAxis = reversedYAxes.find(
    (axis) => axis?.position === "left" || axis?.position === void 0
  );
  const rightAxis = reversedYAxes.find((axis) => axis?.position === "right");
  const sanitizedYAxes = {};
  if (leftAxis) {
    sanitizedYAxes.left = leftAxis;
  }
  if (rightAxis) {
    sanitizedYAxes.right = rightAxis;
  }
  return sanitizedYAxes;
};
function getYAxisConfiguration(slotsYAxis, contextYAxis) {
  const contextConfig = contextYAxis === void 0 ? {} : sanitizeYAxisConfig(contextYAxis);
  const slotsConfig = slotsYAxis === void 0 ? {} : sanitizeYAxisConfig(slotsYAxis);
  const { left, right } = merge({}, contextConfig, slotsConfig);
  if (left === void 0 && right === void 0) {
    return defaultYAxis;
  }
  const result = {};
  if (left) {
    result.left = merge({}, defaultYAxis.left, left);
  }
  if (right) {
    const defaultRight = { ...defaultYAxis.left, position: "right" };
    result.right = merge({}, defaultRight, right);
  }
  return result;
}
function getXAxisConfiguration(contextXAxis, slotsXAxis) {
  return merge({}, defaultXAxis, contextXAxis, slotsXAxis);
}
function getThresholdsConfiguration(contextThresholds = [], slotsThresholds = []) {
  const allThresholds = contextThresholds.concat(slotsThresholds).map((threshold) => ({
    ...threshold,
    id: crypto.randomUUID(),
    position: threshold.position ? threshold.position : "left",
    color: isValidCSSColor(threshold.color) ? threshold.color : DEFAULT_SCALE_FALLBACK_COLOR
  }));
  if (!allThresholds.length) {
    return void 0;
  }
  const thresholdConfig = {};
  thresholdConfig.leftThreshold = allThresholds.filter(
    (threshold) => threshold.position === "left"
  );
  thresholdConfig.rightThreshold = allThresholds.filter(
    (threshold) => threshold.position === "right"
  );
  return thresholdConfig;
}
const getValueRepresentation = (contextValueRepresentation, configValueRepresentation) => contextValueRepresentation ?? configValueRepresentation ?? DEFAULT_VALUE_REPRESENTATION;
function getInfiniteZoomConfiguration(contextConfig, propsConfig) {
  return propsConfig ?? contextConfig ?? DEFAULT_INFINITE_ZOOM;
}
function getInitialZoomConfiguration(contextConfig, propsConfig) {
  return propsConfig ?? contextConfig;
}
function getCurrentZoomConfiguration(contextConfig, propsConfig) {
  return propsConfig ?? contextConfig;
}
function createHistogramConfiguration(slots, contextConfig, histogramChartConfigProps) {
  const colorPalette = getColorPaletteConfiguration(
    contextConfig.colorPalette,
    histogramChartConfigProps?.colorPalette
  );
  const legend = getLegendConfiguration(contextConfig.legend, slots.legend);
  const tooltip = getTooltipConfiguration(contextConfig.tooltip, slots.tooltip);
  const yAxis = getYAxisConfiguration(slots.yAxis, contextConfig.yAxis);
  const xAxis = getXAxisConfiguration(contextConfig.xAxis, slots.xAxis);
  const truncationMode = histogramChartConfigProps?.truncationMode ?? contextConfig?.truncationMode ?? DEFAULT_TRUNCATION_MODE;
  const toolbar = getToolbarConfiguration(
    contextConfig.toolbar,
    slots.toolbar,
    slots.downloadCSV,
    contextConfig.downloadCSV
  );
  const propsInteractions = {
    onZoomChange: histogramChartConfigProps.onZoomChange
  };
  const hasContextInteractions = contextConfig && (contextConfig.pan || contextConfig.zoom);
  const chartConfigInteractions = hasContextInteractions ? {
    pan: { enabled: contextConfig.pan?.disabled !== true },
    zoom: { enabled: contextConfig.zoom?.disabled !== true },
    zoomX: { enabled: contextConfig.zoom?.disabled !== true }
  } : {};
  const interactions = getInteractions(
    contextConfig.interactions,
    slots.interactions ?? {},
    propsInteractions,
    chartConfigInteractions
  );
  const configThresholds = contextConfig.thresholdIndicators || [];
  const thresholds = getThresholdsConfiguration(
    configThresholds,
    slots.thresholds
  );
  const valueRepresentation = getValueRepresentation(
    contextConfig.valueRepresentation,
    histogramChartConfigProps.valueRepresentation
  );
  const initialZoom = getInitialZoomConfiguration(
    contextConfig.initialZoom,
    histogramChartConfigProps.initialZoom
  );
  const currentZoom = getCurrentZoomConfiguration(
    contextConfig.currentZoom,
    histogramChartConfigProps.currentZoom
  );
  const infiniteZoom = getInfiniteZoomConfiguration(
    contextConfig.infiniteZoom,
    histogramChartConfigProps.infiniteZoom
  );
  const colorRules = (contextConfig.colorRules || []).filter(isValidColorRule).concat(slots.colorRules || []);
  return {
    valueRepresentation,
    colorPalette,
    legend,
    tooltip,
    yAxis,
    xAxis,
    truncationMode,
    toolbar,
    annotations: slots.annotations,
    interactions,
    thresholds,
    initialZoom,
    currentZoom,
    infiniteZoom,
    select: slots.select,
    colorRules
  };
}
function deserializeConfig(serializedConfig) {
  try {
    const parsedData = JSON.parse(serializedConfig);
    parsedData.interactions = {
      ...parsedData.interactions,
      zoom: { enabled: parsedData.zoom?.disabled !== true },
      zoomX: { enabled: parsedData.zoom?.disabled !== true }
    };
    return merge({}, defaultHistogramChartConfig, parsedData);
  } catch {
    return defaultHistogramChartConfig;
  }
}
function parseYAxisConfig(yAxis) {
  const { left, right } = yAxis;
  const yAxisConfig = [];
  if (left) {
    yAxisConfig.push(left);
  }
  if (right) {
    yAxisConfig.push(right);
  }
  return yAxisConfig;
}
function parseThresholdsConfig(thresholds) {
  const { leftThreshold, rightThreshold } = thresholds;
  const thresholdsConfig = [];
  if (leftThreshold) {
    thresholdsConfig.push(...leftThreshold);
  }
  if (rightThreshold) {
    thresholdsConfig.push(...rightThreshold);
  }
  return thresholdsConfig;
}
function normalizeConfig(config) {
  const { thresholds, ...rest } = config;
  const thresholdIndicators = thresholds ? parseThresholdsConfig(thresholds) : void 0;
  const { intents, ...toolbar } = config.toolbar ?? {};
  return {
    ...rest,
    toolbar,
    yAxis: parseYAxisConfig(config.yAxis),
    thresholdIndicators
  };
}
export {
  createHistogramConfiguration,
  deserializeConfig,
  getColorPaletteConfiguration,
  getCurrentZoomConfiguration,
  getInfiniteZoomConfiguration,
  getInitialZoomConfiguration,
  getLegendConfiguration,
  getThresholdsConfiguration,
  getTooltipConfiguration,
  getXAxisConfiguration,
  getYAxisConfiguration,
  isValidHistogramTooltipVariant,
  normalizeConfig
};
