// @dynatrace/strato-components/esm/charts/histogram/utils/iterate-slots.js
import { isEmpty, merge } from "lodash-es";
import { Children } from "react";
import { isElement } from "react-is";
import { Intent } from "../../../core/slots/Intent/Intent.js";
import { isSlot } from "../../../core/utils/is-slot.js";
import { populateMissingMarkerProps } from "../../core/components/annotations/utils/populate-missing-marker-props.js";
import { populateMissingTrackProps } from "../../core/components/annotations/utils/populate-missing-track-props.js";
import { ChartInteractions } from "../../core/slots/chart-interactions/ChartInteractions.js";
import { ColorRule, isValidColorRule } from "../../core/slots/ColorRule.js";
import { ChartToolbar } from "../../core/slots/toolbar/ChartToolbar.js";
import { DownloadData } from "../../core/slots/toolbar/DownloadData.js";
import { getConfigPlusDefaultOrEmpty } from "../../core/utils/interactions.js";
import { iterateChartInteractionsSlots } from "../../core/utils/iterate-chart-interactions-slots.js";
import { Annotations } from "../slots/annotations/Annotations.js";
import { Marker } from "../slots/annotations/Marker.js";
import { AnnotationsTooltip } from "../slots/annotations/Tooltip.js";
import { Track } from "../slots/annotations/Track.js";
import { HistogramChartDownloadCSVSlot } from "../slots/DownloadCSV.js";
import { Legend } from "../slots/Legend.js";
import { Pan } from "../slots/Pan.js";
import { Select } from "../slots/Select.js";
import { ThresholdIndicator } from "../slots/ThresholdIndicator.js";
import { Toolbar } from "../slots/Toolbar.js";
import { Tooltip } from "../slots/Tooltip.js";
import { XAxis } from "../slots/XAxis.js";
import { YAxis } from "../slots/YAxis.js";
import { Zoom } from "../slots/Zoom.js";
const slotInteractionsDefaultValues = {
  zoom: { enabled: true },
  zoomX: { enabled: true },
  pan: { enabled: true }
};
const chartInteractionsDefaultValues = {
  zoom: { enabled: false },
  zoomX: { enabled: false },
  pan: { enabled: false }
};
const getInteractionsConfig = (chartConfig, slotsConfig) => {
  const chartApiValues = getConfigPlusDefaultOrEmpty(
    chartInteractionsDefaultValues,
    chartConfig
  );
  const slotsApiValues = getConfigPlusDefaultOrEmpty(
    slotInteractionsDefaultValues,
    slotsConfig
  );
  return merge({}, chartApiValues, slotsApiValues);
};
const disabledOnlyWhenSpecified = ({ disabled }) => disabled;
const iterateSlots = (children) => {
  let slots = {
    legend: {},
    xAxis: {},
    yAxis: [],
    tooltip: {},
    toolbar: {},
    interactions: {},
    thresholds: [],
    colorRules: []
  };
  let slotToolbar = {};
  let chartToolbar = {};
  let chartInteractions = {};
  let slotInteractions = {};
  let intents;
  Children.forEach(children, (child) => {
    if (isSlot(child, ChartInteractions)) {
      const { children: children2, ...previousInteractions } = child.props;
      const interactions = iterateChartInteractionsSlots(children2);
      chartInteractions = {
        ...previousInteractions,
        ...interactions
      };
    } else if (isSlot(child, XAxis)) {
      slots = {
        ...slots,
        xAxis: child.props
      };
    } else if (isSlot(child, YAxis)) {
      slots.yAxis.push(child.props);
    } else if (isSlot(child, Legend)) {
      slots = {
        ...slots,
        legend: child.props
      };
    } else if (isSlot(child, Tooltip)) {
      slots = {
        ...slots,
        tooltip: child.props
      };
    } else if (isSlot(child, ChartToolbar)) {
      const { children: children2, ...toolbar } = child.props;
      const childrenToolbar = iterateChartToolbarSlots(children2);
      chartToolbar = {
        enabled: true,
        ...toolbar,
        ...childrenToolbar
      };
    } else if (isSlot(child, Toolbar)) {
      slotToolbar = {
        enabled: child.props.hidden !== true,
        ...slotToolbar,
        ...child.props
      };
    } else if (isSlot(child, Zoom)) {
      const enabled = !disabledOnlyWhenSpecified(child.props);
      slotInteractions = {
        ...slotInteractions,
        zoom: { enabled },
        zoomX: { enabled }
      };
    } else if (isSlot(child, Pan)) {
      const enabled = !disabledOnlyWhenSpecified(child.props);
      slotInteractions = { ...slotInteractions, pan: { enabled } };
    } else if (isSlot(child, HistogramChartDownloadCSVSlot)) {
      slotToolbar = { ...slotToolbar, downloadData: { enabled: true } };
    } else if (isSlot(child, Intent)) {
      intents = intents ? [...intents, child.props] : [child.props];
    } else if (isSlot(child, Annotations)) {
      const { children: children2, ...annotationsProps } = child.props;
      const { tracks, tooltip } = iterateTrackSlots(children2);
      const annotations = {
        ...annotationsProps,
        tracks,
        tooltip
      };
      slots = {
        ...slots,
        // eslint-disable-next-line @typescript-eslint/prefer-ts-expect-error, @typescript-eslint/ban-ts-comment
        // @ts-ignore Strict mode transition - needs to be investigated
        annotations
      };
    } else if (isSlot(child, ThresholdIndicator)) {
      slots.thresholds.push(child.props);
    } else if (isSlot(child, Select)) {
      slots = {
        ...slots,
        select: child.props
      };
    } else if (isSlot(child, ColorRule)) {
      isValidColorRule(child.props) && slots.colorRules.push(child.props);
    }
  });
  const hasInteractionConfig = Object.keys(chartInteractions).length > 0 || Object.keys(slotInteractions).length > 0;
  slots.interactions = hasInteractionConfig ? getInteractionsConfig(chartInteractions, slotInteractions) : {};
  const hasToolbarConfig = !isEmpty(slotToolbar) || !isEmpty(chartToolbar) || intents;
  if (hasToolbarConfig) {
    slots.toolbar = {
      ...chartToolbar,
      ...slotToolbar,
      enabled: slotToolbar.enabled ?? chartToolbar.enabled ?? true,
      intents
    };
  }
  return slots;
};
const iterateChartToolbarSlots = (children) => {
  let toolbar = {};
  Children.forEach(children, (childToolbar) => {
    if (isSlot(childToolbar, DownloadData)) {
      toolbar = {
        ...toolbar,
        downloadData: { enabled: true }
      };
    }
  });
  return toolbar;
};
const iterateTrackSlots = (children) => {
  const tracks = [];
  let tooltip = {
    template: void 0,
    hidden: false
  };
  Children.forEach(children, (child) => {
    if (isSlot(child, Track)) {
      const { children: children2, ...trackProps } = child.props;
      const currentTrack = {
        ...trackProps,
        markers: []
      };
      Children.forEach(children2, (child2) => {
        if (isElement(child2)) {
          if (isSlot(child2, Marker)) {
            const marker = populateMissingMarkerProps(child2.props, trackProps);
            currentTrack.markers.push(marker);
          }
        }
      });
      const track = populateMissingTrackProps(currentTrack);
      tracks.push(track);
    } else if (isSlot(child, AnnotationsTooltip)) {
      const { children: children2, hidden } = child.props;
      tooltip = {
        template: children2,
        hidden: hidden ?? false
      };
    }
  });
  return { tracks, tooltip };
};
export {
  iterateSlots
};
