// @dynatrace/strato-components/esm/tables/DataTable/components/TableActions/DataTableSelectedRowsActions.js
import { jsx } from "react/jsx-runtime";
import { forwardRef } from "react";
import { useBehavioralTrackingProps } from "../../../../core/hooks/useBehavioralTrackingProps.js";
import { devLog } from "../../../../core/utils/logging.js";
import { Flex } from "../../../../layouts/flex/Flex.js";
import { useDataTable } from "../../contexts/DataTableContext.js";
function _isDataTableSelectedRowsActions(child) {
  return child?.type === DataTableSelectedRowsActions;
}
const DataTableSelectedRowsActions = forwardRef(
  (props, forwardedRef) => {
    const table = useDataTable();
    const {
      children,
      "data-testid": dataTestId,
      className: consumerClassName,
      style: consumerStyle,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
    if (typeof children !== "function") {
      devLog("The SelectedRowsActions sub-component child must be a function!");
      return null;
    }
    const currentSelectableRows = table.getSelectedRowModel().flatRows.map((row) => ({
      rowId: row.id,
      originalRow: row.original
    }));
    return /* @__PURE__ */ jsx(
      Flex,
      {
        ref: forwardedRef,
        justifyContent: "flex-start",
        alignItems: "flex-end",
        flexWrap: "wrap",
        "data-testid": dataTestId,
        className: consumerClassName,
        style: consumerStyle,
        ...behavioralTrackingProps,
        ...restProps,
        children: /* @__PURE__ */ jsx(Flex, { alignItems: "center", flexWrap: "wrap", gap: 4, children: children(currentSelectableRows) })
      }
    );
  }
);
DataTableSelectedRowsActions.displayName = "DataTable.SelectedRowsActions";
export {
  DataTableSelectedRowsActions,
  _isDataTableSelectedRowsActions
};
