// @dynatrace/strato-components/esm/tables/DataTable/components/TableActions/DataTableTableActions.js
import { jsx } from "react/jsx-runtime";
import { forwardRef } from "react";
import { useBehavioralTrackingProps } from "../../../../core/hooks/useBehavioralTrackingProps.js";
import { Flex } from "../../../../layouts/flex/Flex.js";
function _isTableActions(child) {
  return child?.type === DataTableTableActions;
}
const DataTableTableActions = forwardRef(
  (props, forwardedRef) => {
    const {
      children,
      "data-testid": dataTestId,
      className: consumerClassName,
      style: consumerStyle,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
    return /* @__PURE__ */ jsx(
      Flex,
      {
        ref: forwardedRef,
        justifyContent: "flex-start",
        alignItems: "flex-end",
        flexWrap: "wrap",
        gap: 4,
        "data-testid": dataTestId,
        className: consumerClassName,
        style: consumerStyle,
        ...behavioralTrackingProps,
        ...restProps,
        children
      }
    );
  }
);
DataTableTableActions.displayName = "DataTable.TableActions";
export {
  DataTableTableActions,
  _isTableActions
};
