// @dynatrace/strato-components/layouts/shared/resize-handle-styles.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var resize_handle_styles_exports = {};
__export(resize_handle_styles_exports, {
  resizeHandleStyles: () => resizeHandleStyles
});
module.exports = __toCommonJS(resize_handle_styles_exports);
var import_borders = __toESM(require("@dynatrace/strato-design-tokens/borders"));
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
var import_easings = __toESM(require("@dynatrace/strato-design-tokens/easings"));
var import_spacings = __toESM(require("@dynatrace/strato-design-tokens/spacings"));
var import_timings = __toESM(require("@dynatrace/strato-design-tokens/timings"));
var import_visual_refresh_overrides = require("./visual-refresh-overrides.js");
const resizeHandleStyles = {
  backgroundColor: import_colors.default.Border.Neutral.Accent,
  border: "none",
  borderRadius: import_borders.default.Radius.Field.Default,
  cursor: "col-resize",
  height: import_spacings.default.Size36,
  left: "3px",
  margin: 0,
  padding: 0,
  touchAction: "none",
  transition: `height ${import_timings.default.Base.Default} ${import_easings.default.FastInSlowOut}`,
  width: "2px",
  selectors: {
    "&:focus-visible": {
      width: "100%"
    },
    // If we click the handle we don't want to apply the extra handling for the boxShadow
    "&:hover&:focus-visible": {
      width: "2px"
    },
    "&:focus-visible, &:hover": {
      height: import_spacings.default.Size48,
      backgroundColor: import_colors.default.Border.Neutral.AccentHover
    },
    // The resizing ref is applied to the actual handle to manage the focus state by the ref.
    // However, this would make only the handle resizable and not the whole separator, hence the after element.
    "&:after": {
      content: '""',
      display: "block",
      height: "100%",
      left: "50%",
      position: "absolute",
      top: 0,
      transform: "translateX(-50%)",
      userSelect: "none",
      width: "100%"
    },
    ...import_visual_refresh_overrides.vrResizeHandleOverrides
  }
};
