// @dynatrace/strato-components/layouts/shared/visual-refresh-overrides.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var visual_refresh_overrides_exports = {};
__export(visual_refresh_overrides_exports, {
  vrResizeHandleContainerBaseOverrides: () => vrResizeHandleContainerBaseOverrides,
  vrResizeHandleContainerInteractionOverrides: () => vrResizeHandleContainerInteractionOverrides,
  vrResizeHandleContainerOverrides: () => vrResizeHandleContainerOverrides,
  vrResizeHandleFocusOverrides: () => vrResizeHandleFocusOverrides,
  vrResizeHandleOverrides: () => vrResizeHandleOverrides
});
module.exports = __toCommonJS(visual_refresh_overrides_exports);
var import_borders = __toESM(require("@dynatrace/strato-design-tokens/borders"));
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
var import_spacings = __toESM(require("@dynatrace/strato-design-tokens/spacings"));
var import_AppRoot_css = require("../../core/components/app-root/AppRoot.sty.js");
const vrResizeHandleContainerBaseOverrides = {
  backgroundColor: import_colors.default.Background.Base.Default,
  backgroundImage: `linear-gradient(${import_colors.default.Border.Neutral.Subdued}, ${import_colors.default.Border.Neutral.Subdued})`,
  backgroundSize: `${import_borders.default.Width.Default} 100%`,
  backgroundRepeat: "no-repeat",
  backgroundPosition: "center",
  borderRadius: 0,
  height: "100%",
  width: import_spacings.default.Size8,
  left: 0,
  transition: "none"
};
const vrResizeHandleContainerInteractionOverrides = {
  [`.${import_AppRoot_css.dtVisualRefreshEnabled} &:hover`]: {
    backgroundImage: `linear-gradient(${import_colors.default.Border.Neutral.SubduedHover}, ${import_colors.default.Border.Neutral.SubduedHover})`
  },
  [`.${import_AppRoot_css.dtVisualRefreshEnabled} &:active`]: {
    backgroundImage: `linear-gradient(${import_colors.default.Border.Neutral.SubduedActive}, ${import_colors.default.Border.Neutral.SubduedActive})`
  }
};
const vrResizeHandleContainerOverrides = {
  [`.${import_AppRoot_css.dtVisualRefreshEnabled} &`]: vrResizeHandleContainerBaseOverrides,
  ...vrResizeHandleContainerInteractionOverrides
};
const vrResizeHandleOverrides = {
  [`.${import_AppRoot_css.dtVisualRefreshEnabled} &`]: {
    width: 0,
    cursor: "ew-resize"
  }
};
const vrResizeHandleFocusOverrides = {
  backgroundImage: `linear-gradient(${import_colors.default.Border.Neutral.Accent}, ${import_colors.default.Border.Neutral.Accent})`,
  backgroundSize: `${import_borders.default.Width.Emphasized} 100%`
};
