// @dynatrace/strato-components/esm/charts/tree-map/components/canvas/utils/canvas-label-utils.js
import { calculateWidth } from "../../../../core/components/text-ellipsis/truncate-string.js";
import { LABEL_OPTIONS } from "../../../constants.js";
const LINE_SPACING = 2;
const FONT_SIZE_DIFFERENCE = 1;
const getLabelFit = (height, width, fontSize, padding, labelWidth) => {
  const { nameWidth, nameValueWidth, valueWidth, multiLineWidth } = labelWidth;
  const lineSpacing = 2;
  const paddingSum = 2 * padding;
  const availableWidth = width - paddingSum;
  const minInlineHeight = paddingSum + fontSize;
  const minMultiLineHeight = paddingSum + fontSize * 2 + lineSpacing;
  const fitMultiLine = height >= minMultiLineHeight && availableWidth >= multiLineWidth;
  if (fitMultiLine) {
    return LABEL_OPTIONS.Multi;
  }
  const fitInline = height >= minInlineHeight && availableWidth >= nameValueWidth;
  if (fitInline) {
    return LABEL_OPTIONS.Inline;
  }
  const fitName = height >= minInlineHeight && availableWidth >= nameWidth;
  if (fitName) {
    return LABEL_OPTIONS.Name;
  }
  const fitEllipsis = height >= minInlineHeight && availableWidth < nameWidth;
  if (fitEllipsis) {
    const fitAtLeastValue = height >= minMultiLineHeight && availableWidth >= valueWidth;
    return fitAtLeastValue ? LABEL_OPTIONS.EllipsisWithValue : LABEL_OPTIONS.Ellipsis;
  }
  return LABEL_OPTIONS.Fallback;
};
const getRectCenter = (width, height) => ({
  centerX: width / 2,
  centerY: height / 2
});
const getLabelWidth = (nameWidth, value, charWidthMap) => {
  const valueWidth = calculateWidth(value, charWidthMap);
  const nameValueWidth = nameWidth + calculateWidth(` ${value}`, charWidthMap);
  const multiLineWidth = Math.max(nameWidth, valueWidth);
  return { nameWidth, valueWidth, nameValueWidth, multiLineWidth };
};
const getAppliedLabels = (labelFit, name, value) => {
  const labelMap = {
    multi: [name, value],
    inline: [`${name} ${value}`],
    name: [name],
    ellipsis: [name],
    "ellipsis-with-value": [name, value],
    fallback: []
  };
  return labelMap[labelFit];
};
const getOffsetY = (labels, labelIndex, fontSize) => {
  const downOffsetY = Math.round(fontSize / 2) + LINE_SPACING;
  const upOffsetY = -1 * downOffsetY - FONT_SIZE_DIFFERENCE;
  const isMultiLine = labels.length > 1;
  return isMultiLine ? labelIndex === 0 ? upOffsetY : downOffsetY : 0;
};
export {
  FONT_SIZE_DIFFERENCE,
  LINE_SPACING,
  getAppliedLabels,
  getLabelFit,
  getLabelWidth,
  getOffsetY,
  getRectCenter
};
