// @dynatrace/strato-components/filters/segment-selector/parts/EmptySegments.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var EmptySegments_exports = {};
__export(EmptySegments_exports, {
  EmptySegments: () => EmptySegments
});
module.exports = __toCommonJS(EmptySegments_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_Button = require("../../../buttons/button/Button.js");
var import_external_links = require("../../../core/external-links.js");
var import_Flex = require("../../../layouts/flex/Flex.js");
var import_Heading = require("../../../typography/heading/Heading.js");
var import_SegmentsCache = require("../contexts/SegmentsCache/SegmentsCache.js");
var import_SegmentsUIContext = require("../contexts/SegmentsUI/SegmentsUIContext.js");
var import_segmentMessages = require("../segmentMessages.js");
var import_EmptySegments_css = require("./EmptySegments.sty.js");
var import_NoAccessWithoutSelection = require("./noAccessRights/NoAccessWithoutSelection.js");
var styles = __toESM(require("./Overlay.sty.js"));
var import_segmentManagementHref = require("./segmentManagementHref.js");
const EmptySegments = (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const intl = (0, import_react_intl.useIntl)();
    const buttonRef = (0, import_react.useRef)(null);
    const { readOnly } = (0, import_react.useContext)(import_SegmentsUIContext.UISegmentsContext);
    (0, import_react.useEffect)(() => {
      buttonRef.current?.focus();
    }, []);
    if (import_SegmentsCache.segmentsClient.hasAccessRights === false) {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_NoAccessWithoutSelection.NoAccessWithoutSelection, {});
    }
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      import_Flex.Flex,
      {
        flexDirection: "column",
        alignItems: "center",
        padding: 16,
        ref: forwardedRef,
        width: "100vw",
        className: styles.overlayDefaultWidth,
        style: { maxWidth: "320px" },
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ContainerIcon, {}),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Heading.Heading, { level: 6, children: intl.formatMessage(import_segmentMessages.SEGMENTSELECTOR_MESSAGES.noSegmentsConfigured) }),
          /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
            import_Flex.Flex,
            {
              paddingTop: 8,
              width: "100%",
              justifyContent: readOnly ? "center" : "space-between",
              children: [
                !readOnly && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  import_Button.Button,
                  {
                    as: "a",
                    ref: buttonRef,
                    href: (0, import_segmentManagementHref.getSegmentManagementHref)(),
                    target: "_blank",
                    rel: "noopener noreferrer",
                    variant: "emphasized",
                    className: import_EmptySegments_css.AppLinkButtonCSS,
                    children: intl.formatMessage(import_segmentMessages.SEGMENTSELECTOR_MESSAGES.manageSegments)
                  }
                ),
                /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
                  import_Button.Button,
                  {
                    as: "a",
                    href: import_external_links.SEGMENTS_DOCUMENTATION_URL,
                    target: "_blank",
                    rel: "noopener noreferrer",
                    children: [
                      intl.formatMessage(import_segmentMessages.SEGMENTSELECTOR_MESSAGES.docsLinkText),
                      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Suffix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ExternalLinkIcon, {}) })
                    ]
                  }
                )
              ]
            }
          )
        ]
      }
    );
  }
);
