// @dynatrace/strato-components/filters/segment-selector/parts/LoadingSegments.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var LoadingSegments_exports = {};
__export(LoadingSegments_exports, {
  LoadingSegments: () => LoadingSegments
});
module.exports = __toCommonJS(LoadingSegments_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_Skeleton = require("../../../content/skeleton/Skeleton.js");
var import_Flex = require("../../../layouts/flex/Flex.js");
var import_segmentMessages = require("../segmentMessages.js");
var styles = __toESM(require("./Overlay.sty.js"));
const LoadingSegments = (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const intl = (0, import_react_intl.useIntl)();
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      import_Flex.Flex,
      {
        flexDirection: "column",
        alignItems: "center",
        padding: 16,
        ref: forwardedRef,
        width: "100vw",
        className: styles.overlayDefaultWidth,
        "aria-label": intl.formatMessage(
          import_segmentMessages.SEGMENTSELECTOR_MESSAGES.loadingAriaLabel
        ),
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Skeleton.Skeleton, { width: "100%", height: "20px" }),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Skeleton.Skeleton, { width: "100%", height: "20px" }),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Skeleton.Skeleton, { width: "100%", height: "20px" })
        ]
      }
    );
  }
);
