// @dynatrace/strato-components/filters/segment-selector/parts/Overlay.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Overlay_exports = {};
__export(Overlay_exports, {
  SegmentsOverlay: () => SegmentsOverlay
});
module.exports = __toCommonJS(Overlay_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_query = require("@tanstack/react-query");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_segments_presets = require("@dynatrace-sdk/segments-presets");
var import_EmptySegments = require("./EmptySegments.js");
var import_LoadingSegments = require("./LoadingSegments.js");
var import_NoAccessHint = require("./noAccessRights/NoAccessHint.js");
var overlayStyles = __toESM(require("./Overlay.sty.js"));
var import_SegmentFooter = require("./SegmentFooter.js");
var import_SegmentsPresetsList = require("./segmentsPresets/SegmentsPresetsList.js");
var import_SelectionBar = require("./selectionBar/SelectionBar.js");
var import_Button = require("../../../buttons/button/Button.js");
var import_useId = require("../../../core/hooks/useId.js");
var import_FieldSet = require("../../../forms/forms-core/components/field-set/FieldSet.js");
var import_FormFieldMessageContextProvider = require("../../../forms/forms-core/components/form-field/FormFieldMessageContextProvider.js");
var import_useFormFieldMessagesContext = require("../../../forms/forms-core/hooks/useFormFieldMessagesContext.js");
var import_formfield_messages_state_reducer = require("../../../forms/forms-core/state/formfield-messages-state-reducer.js");
var import_Divider = require("../../../layouts/divider/Divider.js");
var import_Flex = require("../../../layouts/flex/Flex.js");
var import_Grid = require("../../../layouts/grid/Grid.js");
var import_SegmentsCache = require("../contexts/SegmentsCache/SegmentsCache.js");
var import_SegmentsUIContext = require("../contexts/SegmentsUI/SegmentsUIContext.js");
var import_SegmentsUIDispatchContext = require("../contexts/SegmentsUI/SegmentsUIDispatchContext.js");
var import_segmentValidationHelper = require("../contexts/SegmentsUI/segmentValidationHelper.js");
var import_useSegments = require("../hooks/useSegments.js");
var import_segmentMessages = require("../segmentMessages.js");
var styles = __toESM(require("../SegmentSelector.sty.js"));
const SegmentsOverlay = ({ labelId }) => {
  const { getAllSegments } = (0, import_SegmentsCache.useSegmentsCache)();
  const { data, isLoading } = (0, import_react_query.useQuery)(
    {
      queryKey: ["getAllSegments"],
      queryFn: () => getAllSegments().catch(() => /* @__PURE__ */ new Map()),
      notifyOnChangeProps: ["dataUpdatedAt"],
      throwOnError: false
    },
    import_SegmentsCache.segmentsClient.queryClient
  );
  const { isLoading: isLoadingPresets } = (0, import_react_query.useQuery)(
    {
      queryKey: ["getSegmentsPresets"],
      queryFn: () => (
        // We don't want to render the error fallback in case the presets are not available.
        // However, we sill want to keep the loading state until both queries are resolved.
        (0, import_segments_presets.getSegmentsPresets)().catch(() => {
          return {
            pinnedPresets: [],
            recentPresets: []
          };
        })
      )
    },
    import_SegmentsCache.segmentsClient.queryClient
  );
  if (isLoading || isLoadingPresets) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_LoadingSegments.LoadingSegments, {});
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(SegmentsContent, { cachedSegments: data ?? /* @__PURE__ */ new Map(), id: labelId });
};
async function applySegments(uiSegments, setSegments, overlayState) {
  try {
    const applicableSegments = await Promise.all(
      uiSegments.filter((uiSegment) => !uiSegment.isPlaceholder).map(async (uiSegment) => {
        const errorKey = await (0, import_segmentValidationHelper.getErrorKey)(uiSegment);
        if (errorKey === "needs-update") {
          return {
            id: uiSegment.id
          };
        }
        return {
          id: uiSegment.id,
          variables: uiSegment.variables
        };
      })
    );
    await setSegments(applicableSegments);
    const isValid = await (0, import_segmentValidationHelper.areUiSegmentsValid)(applicableSegments);
    if (isValid) {
      overlayState?.close();
    }
  } catch {
  }
}
const SegmentsContent = ({ cachedSegments, id }) => {
  const { dispatch, overlayState } = (0, import_react.useContext)(import_SegmentsUIDispatchContext.UISegmentsDispatchContext);
  const { dispatch: parentErrorDispatch } = (0, import_useFormFieldMessagesContext.useFormFieldMessagesContext)();
  const intl = (0, import_react_intl.useIntl)();
  const addButtonRef = (0, import_react.useRef)(null);
  const {
    uiSegments,
    hasSelection,
    readOnly,
    focusedSegmentId,
    arrayIdentifierErrorIds
  } = (0, import_react.useContext)(import_SegmentsUIContext.UISegmentsContext);
  const formId = (0, import_useId.useId)();
  const formRef = (0, import_react.useRef)(null);
  (0, import_react.useEffect)(() => {
    const errorIds = arrayIdentifierErrorIds;
    return () => {
      if (!errorIds) {
        return;
      }
      if (errorIds.length > 0) {
        parentErrorDispatch?.({
          type: import_formfield_messages_state_reducer.FormFieldMessagesAction.ADDMESSAGE,
          payload: {
            id: "array-identifier",
            validityStateKey: "customError",
            variant: "error",
            messageType: "formcontrol",
            message: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_react_intl.FormattedMessage,
              {
                ...import_segmentMessages.SEGMENTSELECTOR_MESSAGES.arrayIdentifierError
              }
            )
          }
        });
        return;
      }
      parentErrorDispatch?.({
        type: import_formfield_messages_state_reducer.FormFieldMessagesAction.REMOVEMESSAGE,
        payload: {
          id: "array-identifier",
          messageType: "formcontrol"
        }
      });
    };
  }, [arrayIdentifierErrorIds]);
  const availableSegments = (0, import_react.useMemo)(() => {
    const tmp = [];
    cachedSegments.forEach((value, key) => {
      if (!value.temporarilyApplied && !uiSegments.some((segment) => segment.id === key)) {
        tmp.push({ ...value.segment });
      }
    });
    return tmp.sort(
      (firstSegment, secondSegment) => firstSegment.name.localeCompare(secondSegment.name)
    );
  }, [cachedSegments, uiSegments]);
  const {
    segments,
    removeAllSegments: removeAllSegmentsFromStore,
    setSegments
  } = (0, import_useSegments.useSegments)();
  const { removeAllUISegments, clearPlaceholder } = (0, import_SegmentsUIContext.useSegmentsUIContext)();
  if (segments.length === 0 && cachedSegments.size === 0) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_EmptySegments.EmptySegments, {});
  }
  const MessageWrapper = import_SegmentsCache.segmentsClient.hasAccessRights ? import_FormFieldMessageContextProvider.FormFieldMessageContextProvider : import_react.Fragment;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_Flex.Flex,
    {
      px: 4,
      pb: 8,
      gap: 0,
      flexDirection: "column",
      id,
      width: "100vw",
      className: overlayStyles.overlayDefaultWidth,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { className: overlayStyles.scrollContainer, children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
            import_Grid.Grid,
            {
              as: "form",
              ref: formRef,
              id: formId,
              px: 12,
              pt: 12,
              pb: 16,
              onInvalid: (event) => {
                event.stopPropagation();
                if (event.nativeEvent.target.validity.customError && !event.nativeEvent.target.validity.valueMissing) {
                  applySegments(uiSegments, setSegments, overlayState).catch(
                    () => void 0
                  );
                }
              },
              onSubmit: (event) => {
                event.preventDefault();
                event.stopPropagation();
                applySegments(uiSegments, setSegments, overlayState).catch(
                  () => void 0
                );
              },
              onKeyDown: (event) => {
                if (event.key === "Enter" && (event.ctrlKey || event.metaKey)) {
                  formRef?.current?.requestSubmit();
                }
              },
              children: [
                /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_FieldSet.FieldSet, { className: overlayStyles.fieldSet, children: [
                  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_FieldSet.FieldSet.Legend, { className: overlayStyles.legend, children: intl.formatMessage(import_segmentMessages.SEGMENTSELECTOR_MESSAGES.legend) }),
                  /* @__PURE__ */ (0, import_jsx_runtime.jsx)("ul", { className: styles.segmentSelectorList, children: uiSegments?.map((segment) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(MessageWrapper, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                    import_SelectionBar.SelectionBar,
                    {
                      uiSegment: segment,
                      segment: cachedSegments.get(segment.id),
                      readOnly,
                      availableSegments,
                      clearable: segment.isPlaceholder ? uiSegments.length > 1 : true,
                      autoFocus: segment.id === focusedSegmentId,
                      returnFocusTo: addButtonRef
                    },
                    segment.id
                  ) }, segment.id)) }),
                  !readOnly && hasSelection && import_SegmentsCache.segmentsClient.hasAccessRights !== false && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Grid.Grid, { gridTemplateColumns: "1fr auto auto", children: [
                    /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
                      import_Button.Button,
                      {
                        onClick: () => dispatch({
                          type: "add",
                          payload: { isPlaceholder: true }
                        }),
                        variant: "emphasized",
                        ref: addButtonRef,
                        "data-dt-name": "Add",
                        children: [
                          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                            import_Button.Button.Prefix,
                            {
                              "aria-label": intl.formatMessage(
                                import_segmentMessages.SEGMENTSELECTOR_MESSAGES.addSegment
                              ),
                              children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.PlusIcon, {})
                            }
                          ),
                          "Segment"
                        ]
                      }
                    ),
                    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                      import_Button.Button,
                      {
                        onClick: () => {
                          removeAllSegmentsFromStore();
                          removeAllUISegments();
                          overlayState?.close();
                        },
                        type: "reset",
                        form: formId,
                        "data-dt-name": "ClearAll",
                        children: intl.formatMessage(import_segmentMessages.SEGMENTSELECTOR_MESSAGES.clearAll)
                      }
                    ),
                    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                      import_Button.Button,
                      {
                        type: "submit",
                        form: formId,
                        color: "primary",
                        variant: "accent",
                        "data-dt-name": "Apply",
                        onClick: () => clearPlaceholder(),
                        children: intl.formatMessage(import_segmentMessages.SEGMENTSELECTOR_MESSAGES.applyButton)
                      }
                    )
                  ] })
                ] }),
                import_SegmentsCache.segmentsClient.hasAccessRights === false && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_NoAccessHint.NoAccessHint, { readOnly })
              ]
            }
          ),
          !readOnly && import_SegmentsCache.segmentsClient.hasAccessRights === true && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_SegmentsPresetsList.SegmentsPresetList, { showDivider: hasSelection })
        ] }),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Divider.Divider, {}),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_SegmentFooter.SegmentFooter, {})
      ]
    }
  );
};
