// @dynatrace/strato-components/filters/segment-selector/parts/Overlay.sty.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Overlay_css_exports = {};
__export(Overlay_css_exports, {
  divider: () => divider,
  fieldSet: () => fieldSet,
  isAnimating: () => isAnimating,
  legend: () => legend,
  link: () => link,
  menuLink: () => menuLink,
  overlayDefaultWidth: () => overlayDefaultWidth,
  scrollContainer: () => scrollContainer
});
module.exports = __toCommonJS(Overlay_css_exports);
var import_Overlay_css_ts_vanilla = require("./Overlay.css");
var divider = "_bqi6d90-3-6-0";
var fieldSet = "_bqi6d97-3-6-0";
var isAnimating = "_bqi6d94-3-6-0";
var legend = "_bqi6d96-3-6-0";
var link = "_bqi6d95-3-6-0";
var menuLink = "_bqi6d91-3-6-0";
var overlayDefaultWidth = "_bqi6d92-3-6-0";
var scrollContainer = "_bqi6d93-3-6-0";
