// @dynatrace/strato-components/filters/segment-selector/parts/SegmentFooter.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SegmentFooter_exports = {};
__export(SegmentFooter_exports, {
  SegmentFooter: () => SegmentFooter
});
module.exports = __toCommonJS(SegmentFooter_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_segmentManagementHref = require("./segmentManagementHref.js");
var import_segmentMessages = require("../segmentMessages.js");
var styles = __toESM(require("./Overlay.sty.js"));
var import_Button = require("../../../buttons/button/Button.js");
var import_external_links = require("../../../core/external-links.js");
var import_useOverlayStacking = require("../../../core/hooks/useOverlayStacking.js");
var import_Grid = require("../../../layouts/grid/Grid.js");
var import_Tooltip = require("../../../overlays/tooltip/Tooltip.js");
var import_ExternalLink = require("../../../typography/external-link/ExternalLink.js");
var import_SegmentsCache = require("../contexts/SegmentsCache/SegmentsCache.js");
var import_SegmentsUIContext = require("../contexts/SegmentsUI/SegmentsUIContext.js");
const SegmentFooter = () => {
  const intl = (0, import_react_intl.useIntl)();
  const stackingContext = (0, import_useOverlayStacking.useOverlayStacking)(void 0, "overlay");
  const { readOnly } = (0, import_react.useContext)(import_SegmentsUIContext.UISegmentsContext);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_Grid.Grid,
    {
      rowGap: 8,
      columnGap: 0,
      px: 12,
      pb: import_SegmentsCache.segmentsClient.hasAccessRights ? 2 : 4,
      pt: 2,
      mt: 8,
      gridTemplateColumns: import_SegmentsCache.segmentsClient.hasAccessRights && !readOnly ? "auto 1fr" : "1fr",
      justifyItems: "end",
      alignItems: "center",
      children: [
        import_SegmentsCache.segmentsClient.hasAccessRights && !readOnly && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_Tooltip.Tooltip,
          {
            placement: "right",
            text: intl.formatMessage(import_segmentMessages.SEGMENTSELECTOR_MESSAGES.manageSegments),
            style: { zIndex: Number(stackingContext.style.zIndex) + 1 },
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_Button.Button,
              {
                as: "a",
                href: (0, import_segmentManagementHref.getSegmentManagementHref)(),
                target: "_blank",
                rel: "noopener noreferrer",
                size: "condensed",
                className: styles.link,
                "aria-label": intl.formatMessage(
                  import_segmentMessages.SEGMENTSELECTOR_MESSAGES.manageSegments
                ),
                "data-dt-name": "ManageSegments",
                children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.SettingIcon, {})
              }
            )
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_ExternalLink.ExternalLink,
          {
            href: import_external_links.SEGMENTS_DOCUMENTATION_URL,
            className: styles.link,
            "data-dt-name": "LearnMoreAboutSegments",
            children: intl.formatMessage(import_segmentMessages.SEGMENTSELECTOR_MESSAGES.docsLinkText)
          }
        )
      ]
    }
  );
};
