// @dynatrace/strato-components/filters/segment-selector/parts/TooltipMessages.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TooltipMessages_exports = {};
__export(TooltipMessages_exports, {
  TooltipMessages: () => TooltipMessages
});
module.exports = __toCommonJS(TooltipMessages_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_query = require("@tanstack/react-query");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_SkeletonText = require("../../../content/skeleton/SkeletonText.js");
var import_Flex = require("../../../layouts/flex/Flex.js");
var import_Text = require("../../../typography/text/Text.js");
var import_SegmentsCache = require("../contexts/SegmentsCache/SegmentsCache.js");
var import_useSegments = require("../hooks/useSegments.js");
var import_segmentMessages = require("../segmentMessages.js");
var styles = __toESM(require("../SegmentSelector.sty.js"));
var import_utils = require("../utils.js");
const SegmentWithVariablesText = ({
  id,
  name,
  variables
}) => {
  const intl = (0, import_react_intl.useIntl)();
  const endOfLine = variables && variables.length > 1 ? ";" : "";
  const values = variables?.map((value) => (0, import_utils.getReadableValue)(value, intl)).join(", ") + endOfLine;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { style: { lineHeight: 1 }, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Text.Text, { textStyle: "small-emphasized", children: [
      name,
      ": "
    ] }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { textStyle: "small", children: values })
  ] }, id);
};
const WrappedTooltipMessages = ({ ariaLabel }) => {
  const { segments } = (0, import_useSegments.useSegments)();
  const { getSegment } = (0, import_SegmentsCache.useSegmentsCache)();
  const result = (0, import_react_query.useSuspenseQueries)(
    {
      queries: segments.map((segment) => ({
        queryKey: ["getSegment", segment.id],
        queryFn: () => getSegment(segment.id),
        select: (data) => ({
          selectedSegment: segment,
          fetchedSegment: data?.segment
        }),
        retryOnMount: false
      }))
    },
    import_SegmentsCache.segmentsClient.queryClient
  );
  if (segments.length === 0) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { textStyle: "small-emphasized", className: styles.tooltipMaxHeight, children: ariaLabel ?? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_intl.FormattedMessage, { ...import_segmentMessages.SEGMENTSELECTOR_MESSAGES.defaultAriaLabel }) });
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Flex.Flex, { flexDirection: "column", gap: 4, className: styles.tooltipMaxHeight, children: result.map((query) => {
    const selectedSegment = query.data?.selectedSegment;
    const id = selectedSegment.id;
    const name = query.data?.fetchedSegment?.name ?? id;
    if (selectedSegment?.variables) {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        SegmentWithVariablesText,
        {
          id,
          name,
          variables: selectedSegment.variables?.[0]?.values
        },
        id
      );
    }
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { textStyle: "small-emphasized", children: name }, id);
  }) });
};
const TooltipMessages = ({ ariaLabel }) => (
  /** Due to display: 'contents' on the tooltip we need to define a fixed width here. */
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react.Suspense, { fallback: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_SkeletonText.SkeletonText, { width: "150px" }), children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(WrappedTooltipMessages, { ariaLabel }) })
);
