// @dynatrace/strato-components/tables/DataTable/features/Pagination/PaginationInternal.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var PaginationInternal_exports = {};
__export(PaginationInternal_exports, {
  DataTablePaginationInternal: () => DataTablePaginationInternal
});
module.exports = __toCommonJS(PaginationInternal_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_intl = require("react-intl");
var import_breakpoints_js = __toESM(require("@dynatrace/strato-design-tokens/breakpoints-js"));
var import_strato_icons = require("@dynatrace/strato-icons");
var import_Button = require("../../../../buttons/button/Button.js");
var import_useId = require("../../../../core/hooks/useId.js");
var import_Select = require("../../../../forms/select/Select.js");
var import_Flex = require("../../../../layouts/flex/Flex.js");
var import_useBreakpoint = require("../../../../layouts/hooks/useBreakpoint.js");
var import_Text = require("../../../../typography/text/Text.js");
var import_DataTableContext = require("../../contexts/DataTableContext.js");
var import_messages = require("../../messages.js");
const PaginationNavigation = ({
  pageSize,
  pageCount,
  pageIndex,
  isPrevPageDisabled,
  isNextPageDisabled,
  isServerSidePaginated,
  totalRowsCount,
  onPageIndexChange
}) => {
  const pageIndexSelectId = (0, import_useId.useId)();
  const intl = (0, import_react_intl.useIntl)();
  const smallScreen = (0, import_useBreakpoint.useBreakpoint)(`(max-width: ${import_breakpoints_js.default.Max.Mobile}px)`);
  const totalPages = isServerSidePaginated && totalRowsCount !== void 0 ? Math.ceil(totalRowsCount / pageSize) : pageCount;
  const pageCountArray = Array.from(
    { length: totalPages },
    (_, i) => i
  );
  const hasMultiplePages = totalPages > 1;
  const shouldShowPageInfo = !isServerSidePaginated || isServerSidePaginated && totalPages !== -1;
  const totalPagesCountText = hasMultiplePages ? intl.formatMessage(import_messages.dataTableMessages.paginationTotalPages, { totalPages }) : intl.formatMessage(import_messages.dataTableMessages.paginationSinglePageInfo);
  const handlePageIndexSelect = (newPageIndex) => newPageIndex !== null && onPageIndexChange(newPageIndex, "user");
  const handlePreviousPage = () => {
    const prevPageIndex = Math.max(0, pageIndex - 1);
    onPageIndexChange(prevPageIndex, "user");
  };
  const handleNextPage = () => onPageIndexChange(pageIndex + 1, "user");
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { gap: 16, justifySelf: "flex-end", children: [
    shouldShowPageInfo && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { flexWrap: "nowrap", alignItems: "center", children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { as: "p", children: intl.formatMessage(import_messages.dataTableMessages.paginationPageSelection) }),
      hasMultiplePages && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_Select.Select,
        {
          id: pageIndexSelectId,
          value: pageIndex,
          onChange: handlePageIndexSelect,
          "aria-label": intl.formatMessage(import_messages.dataTableMessages.selectPage),
          "data-dt-name": "PaginationPageIndex",
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Select.Select.Content, { children: pageCountArray.length === 0 ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Select.Select.Option, { id: 0, value: 0, children: 0 }, "pageIndex-0") : pageCountArray.map((pageIndexOption) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_Select.Select.Option,
            {
              id: pageIndexOption,
              value: pageIndexOption,
              children: pageIndexOption + 1
            },
            `pageIndex-${pageIndexOption}`
          )) })
        }
      ),
      (!hasMultiplePages || !smallScreen) && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { as: "span", textStyle: "base", style: { whiteSpace: "nowrap" }, children: totalPagesCountText })
    ] }),
    (!smallScreen || isServerSidePaginated) && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { gap: 2, children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_Button.Button,
        {
          onClick: handlePreviousPage,
          disabled: isPrevPageDisabled,
          "aria-label": intl.formatMessage(
            import_messages.dataTableMessages.paginationPrevPage
          ),
          variant: "default",
          "data-dt-name": "PaginationPreviousPage",
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ChevronLeftSmallIcon, {}) })
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_Button.Button,
        {
          onClick: handleNextPage,
          disabled: isNextPageDisabled,
          "aria-label": intl.formatMessage(
            import_messages.dataTableMessages.paginationNextPage
          ),
          variant: "default",
          "data-dt-name": "PaginationNextPage",
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ChevronRightSmallIcon, {}) })
        }
      )
    ] })
  ] });
};
function DataTablePaginationInternal(props) {
  const table = (0, import_DataTableContext.useDataTable)();
  const intl = (0, import_react_intl.useIntl)();
  const pageSizeSelectId = (0, import_useId.useId)();
  const paginationState = table.getState().pagination;
  const pageSize = paginationState.pageSize;
  const onPageSizeChange = table.setPageSize;
  const pageSizeOptions = table.options.availablePaginationOptions ?? [];
  const hasPageSizeOptions = pageSizeOptions.length > 1;
  const { manualPagination } = table.options;
  const { enablePreviousPage, enableNextPage } = paginationState;
  const isServerSidePaginated = enableNextPage !== void 0 || enablePreviousPage !== void 0;
  const isPrevPageDisabled = manualPagination ? !enablePreviousPage : !table.getCanPreviousPage();
  const isNextPageDisabled = manualPagination ? !enableNextPage : !table.getCanNextPage();
  const {
    id,
    "data-testid": dataTestId,
    className: consumerClassName,
    style: consumerStyle,
    ...remainingProps
  } = props;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_Flex.Flex,
    {
      id,
      "data-testid": dataTestId,
      className: consumerClassName,
      style: consumerStyle,
      justifyContent: hasPageSizeOptions ? "space-between" : "flex-end",
      paddingTop: 8,
      role: "navigation",
      flexWrap: "wrap",
      ...remainingProps,
      children: [
        hasPageSizeOptions && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { alignItems: "center", children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_Select.Select,
            {
              id: pageSizeSelectId,
              value: pageSize,
              required: true,
              onChange: (newPageSize) => newPageSize !== null && onPageSizeChange?.(newPageSize),
              "aria-label": intl.formatMessage(import_messages.dataTableMessages.selectPageSize),
              "data-dt-name": "PaginationPageSize",
              children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Select.Select.Content, { children: pageSizeOptions.map((pageSize2) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_Select.Select.Option,
                {
                  id: pageSize2,
                  value: pageSize2,
                  children: `${pageSize2}`
                },
                `${pageSize2}`
              )) })
            }
          ),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { as: "span", style: { whiteSpace: "nowrap" }, children: intl.formatMessage(import_messages.dataTableMessages.paginationRowsPerPage) })
        ] }),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          PaginationNavigation,
          {
            pageSize,
            pageCount: table.getPageCount(),
            pageIndex: paginationState.pageIndex,
            isPrevPageDisabled,
            isNextPageDisabled,
            isServerSidePaginated,
            totalRowsCount: paginationState.totalRowsCount,
            onPageIndexChange: table.setPageIndex
          }
        )
      ]
    }
  );
}
