// @dynatrace/strato-components/tables/DataTable/features/Pagination/Pagination.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Pagination_exports = {};
__export(Pagination_exports, {
  DataTablePagination: () => DataTablePagination,
  DataTablePaginationFeature: () => DataTablePaginationFeature,
  _isDataTablePagination: () => _isDataTablePagination,
  usePagination: () => usePagination
});
module.exports = __toCommonJS(Pagination_exports);
var import_react_table = require("@tanstack/react-table");
var import_react = require("react");
var import_react_is = require("react-is");
var import_pagination_row_model = require("./pagination-row-model.js");
var import_useControllableState = require("../../../../core/hooks/useControllableState.js");
var import_useDataTableConfig = require("../../../hooks/useDataTableConfig.js");
var import_constants = require("../../constants.js");
function DataTablePagination(props) {
  return null;
}
function _isDataTablePagination(child) {
  return (0, import_react_is.isElement)(child) && child.type === DataTablePagination;
}
function getDefaultPageSize(pageSizeOptions, defaultPageSize) {
  if (typeof defaultPageSize === "number") {
    return defaultPageSize;
  }
  if (Array.isArray(pageSizeOptions) && pageSizeOptions[0]) {
    return pageSizeOptions[0];
  }
  return import_constants.DATA_TABLE_DEFAULT_PAGE_SIZE_OPTION;
}
const DataTablePaginationFeature = {
  createTable(table) {
    table.setPagination = (updater, trigger) => {
      const safeUpdater = (old) => {
        const newState = (0, import_react_table.functionalUpdate)(
          updater,
          old
        );
        return newState;
      };
      return table.options.onPaginationChange?.(safeUpdater, trigger);
    };
    table.setPageIndex = (updater, trigger) => {
      table.setPagination((old) => {
        let pageIndex = (0, import_react_table.functionalUpdate)(updater, old.pageIndex);
        const maxPageIndex = typeof table.options.pageCount === "undefined" || table.options.pageCount === -1 ? Number.MAX_SAFE_INTEGER : table.options.pageCount - 1;
        pageIndex = Math.max(0, Math.min(pageIndex, maxPageIndex));
        return {
          ...old,
          pageIndex
        };
      }, trigger);
    };
    table.resetPageIndex = (defaultState) => {
      table.setPageIndex(
        defaultState ? import_constants.DATA_TABLE_DEFAULT_PAGINATION_PAGE_INDEX : table.initialState?.pagination?.pageIndex ?? import_constants.DATA_TABLE_DEFAULT_PAGINATION_PAGE_INDEX,
        "internal"
      );
    };
    table.getPrePaginationRowModel = () => table.getRowOrderRowModel();
  }
};
function usePagination(props, options) {
  const configPagination = (0, import_useDataTableConfig.useDataTableConfigElement)("pagination");
  const paginationElement = import_react.Children.toArray(props.children).find(
    _isDataTablePagination
  );
  options.enablePagination = false;
  const paginationSizeOptions = paginationElement?.props.pageSizeOptions ?? [
    ...import_constants.DATA_TABLE_DEFAULT_PAGE_SIZE_OPTIONS
  ];
  const defaultPaginationSizeOption = getDefaultPageSize(
    paginationElement?.props.pageSizeOptions,
    paginationElement?.props.defaultPageSize ?? configPagination?.defaultPageSize
  );
  const [pageSize, setPageSize] = (0, import_useControllableState.useControllableState)({
    value: paginationElement?.props.pageSize,
    defaultValue: defaultPaginationSizeOption,
    onChange: paginationElement?.props.onPageSizeChange
  });
  const defaultPaginationIndexOption = paginationElement?.props.defaultPageIndex ?? configPagination?.defaultPageIndex ?? import_constants.DATA_TABLE_DEFAULT_PAGINATION_PAGE_INDEX;
  const [pageIndex, setPageIndex] = (0, import_useControllableState.useControllableState)({
    value: paginationElement?.props.pageIndex,
    defaultValue: defaultPaginationIndexOption
  });
  if (!paginationElement) {
    return;
  }
  options.enablePagination = true;
  options.availablePaginationOptions = paginationSizeOptions;
  options.controlledPageIndex = paginationElement.props.pageIndex !== void 0;
  if (paginationElement.props.autoResetPageIndex === false) {
    options.autoResetPageIndex = paginationElement.props.autoResetPageIndex;
  }
  if (paginationElement.props.enableNextPage !== void 0 && paginationElement.props.enablePreviousPage !== void 0) {
    options.manualPagination = true;
    options.pageCount = -1;
  } else {
    options.manualPagination = false;
    options.getPaginationRowModel = (0, import_pagination_row_model.getPaginationRowModel)();
  }
  options.state = {
    ...options.state,
    pagination: {
      pageSize,
      pageIndex,
      enablePreviousPage: paginationElement.props.enablePreviousPage,
      enableNextPage: paginationElement.props.enableNextPage,
      totalRowsCount: paginationElement.props.totalRowsCount
    }
  };
  options.onPaginationChange = (updater, trigger = "user") => {
    const nextState = (0, import_react_table.functionalUpdate)(
      updater,
      options.state?.pagination ?? {
        pageSize,
        pageIndex
      }
    );
    if (nextState?.pageSize !== pageSize && nextState?.pageSize !== void 0) {
      setPageSize(nextState.pageSize);
      setPageIndex(import_constants.DATA_TABLE_DEFAULT_PAGINATION_PAGE_INDEX);
      paginationElement?.props.onPageIndexChange?.(
        import_constants.DATA_TABLE_DEFAULT_PAGINATION_PAGE_INDEX,
        "internal"
      );
    } else if (nextState?.pageIndex !== pageIndex && nextState?.pageIndex !== void 0) {
      setPageIndex(nextState.pageIndex);
      paginationElement?.props.onPageIndexChange?.(
        nextState.pageIndex,
        trigger
      );
    }
  };
}
