// @dynatrace/strato-components/charts/timeseries/hooks/useBuildFilteredMetadata.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useBuildFilteredMetadata_exports = {};
__export(useBuildFilteredMetadata_exports, {
  useBuildFilteredMetadata: () => useBuildFilteredMetadata
});
module.exports = __toCommonJS(useBuildFilteredMetadata_exports);
var import_lodash_es = require("lodash");
var import_useMinMaxXAxis = require("./useMinMaxXAxis.js");
var import_useXAxisConfig = require("./useXAxisConfig.js");
var import_useYAxisConfig = require("./useYAxisConfig.js");
var import_apply_gap_policy = require("../../core/utils/apply-gap-policy.js");
var import_get_data_point_is_gap = require("../../core/utils/get-data-point-is-gap.js");
var import_safe_get = require("../../core/utils/safe-get.js");
var import_datapoint_stacking = require("../components/displays/shared/utils/datapoint-stacking.js");
var import_get_datapoint_value = require("../utils/get-datapoint-value.js");
var import_get_relative_conf = require("../utils/get-relative-conf.js");
var import_get_sanitized_value_representation = require("../utils/get-sanitized-value-representation.js");
var import_is_stacked = require("../utils/is-stacked.js");
var import_timeseries3 = require("../utils/timeseries.js");
function useBuildFilteredMetadata(data, valueRepPerVariant, seriesMap, units) {
  const xAxisConfig = (0, import_useXAxisConfig.useXAxisConfig)();
  const yAxisConfig = (0, import_useYAxisConfig.useYAxisConfig)();
  const { min, max } = (0, import_useMinMaxXAxis.useMinMaxXAxis)();
  const datapointsMap = /* @__PURE__ */ new WeakMap();
  const sanitizedValueRepPerVariant = (0, import_get_sanitized_value_representation.getSanitizedValueRepresentation)(
    data,
    valueRepPerVariant
  );
  const filteredTimeseries = { ...data };
  const filteredUnits = [];
  const originalFilteredUnits = [];
  let relativeGroupedValues;
  for (const variant of data.shapePriority) {
    const cumulativeStackedPoints = /* @__PURE__ */ new Map();
    const lastSignStacked = /* @__PURE__ */ new Map();
    const isBand = variant === "band";
    const variantSeries = data[variant];
    relativeGroupedValues = (0, import_get_relative_conf.getRelativeConf)(
      variantSeries,
      variant,
      sanitizedValueRepPerVariant,
      !(0, import_is_stacked.isStacked)(variant) ? relativeGroupedValues : void 0
    );
    const newSeries = [];
    for (let i = 0; i < variantSeries.length; i++) {
      const series = variantSeries[i];
      const seriesMetadata = (0, import_safe_get.safeGet)(seriesMap, series);
      const { id, color, gapPolicy, pointsDisplay } = seriesMetadata;
      const unit = relativeGroupedValues ? "%" : series.unit;
      const originalUnit = series.unit;
      const metadata = {
        id,
        color,
        unit,
        originalUnit,
        gapPolicy,
        pointsDisplay
      };
      if (units.length === 1 && yAxisConfig.right && !yAxisConfig.left) {
        filteredUnits.push(void 0);
      }
      filteredUnits.push(unit);
      originalFilteredUnits.push(originalUnit);
      seriesMap.set(series, metadata);
      const thresholdGap = (0, import_apply_gap_policy.getThresholdValue)(gapPolicy);
      const newDatapoints = [];
      for (let j = 0; j < series.datapoints.length; j++) {
        const datapoint = series.datapoints[j];
        if (xAxisConfig && (xAxisConfig.min || xAxisConfig.max)) {
          const dataPointTimestamp = (0, import_timeseries3.getDatapointTime)(
            datapoint.start,
            datapoint.end,
            datapoint.center
          );
          const dataPointOutOfBounds = dataPointTimestamp < +min || dataPointTimestamp > +max;
          if (dataPointOutOfBounds) {
            continue;
          }
        }
        const nextDatapoint = series.datapoints[j + 1];
        const { absolute, value, y0, absoluteY0, y1, absoluteY1 } = (0, import_get_datapoint_value.getDatapointValue)(
          datapoint,
          variant,
          originalUnit,
          relativeGroupedValues
        );
        const isBandPointUndefined = isBand && ((0, import_lodash_es.isNil)(y0) || (0, import_lodash_es.isNil)(y1));
        const isPointUndefined = !isBand && (0, import_lodash_es.isNil)(value);
        if (isBandPointUndefined || isPointUndefined) {
          continue;
        }
        const isGap = gapPolicy !== "connect" && (0, import_get_data_point_is_gap.getDatapointIsGap)(datapoint, nextDatapoint, thresholdGap);
        let stackedSign = value ?? 0;
        const key = datapoint.start.getTime().toString() + datapoint.end?.getTime();
        if (value === 0) {
          const prevSign = lastSignStacked.get(key) ?? 0;
          stackedSign = prevSign;
        }
        const cumulativeStack = (0, import_is_stacked.isStacked)(variant) ? (0, import_datapoint_stacking.getCumulativeValue)(
          cumulativeStackedPoints,
          datapoint,
          relativeGroupedValues ? originalUnit : unit,
          id,
          value,
          stackedSign
        ) : void 0;
        const datapointMetadata = {
          start: datapoint.start,
          end: datapoint.end,
          center: datapoint.center,
          seriesId: id,
          color,
          unit,
          isGap: false,
          absoluteValue: absolute,
          value,
          variant,
          y0,
          y1,
          absoluteY0,
          absoluteY1,
          cumulativeValue: cumulativeStack?.cumulativeValue ?? 0,
          seriesName: series.name,
          originalUnit
        };
        if ((0, import_is_stacked.isStacked)(variant) && cumulativeStack) {
          cumulativeStackedPoints.set(cumulativeStack.key, {
            seriesId: id,
            datapointsCumulative: cumulativeStack.datapointsCumulative,
            baseCumulative: cumulativeStack.baseCumulative
          });
          lastSignStacked.set(key, stackedSign);
        }
        datapointsMap.set(datapoint, datapointMetadata);
        newDatapoints.push(datapoint);
        if (isGap) {
          const metadataGappedDatapoint = {
            ...datapointMetadata,
            isGap: true,
            value: Infinity
          };
          const gappedDatapoint = {
            start: datapoint.end ?? datapoint.start,
            end: nextDatapoint.start,
            isGap: true,
            value: Infinity
          };
          datapointsMap.set(gappedDatapoint, metadataGappedDatapoint);
          newDatapoints.push(gappedDatapoint);
        }
      }
      const { datapoints, ...spreadSeries } = series;
      const newSingleSeries = { ...spreadSeries, datapoints: newDatapoints };
      newSeries.push(newSingleSeries);
      seriesMap.set(newSingleSeries, metadata);
    }
    filteredTimeseries[variant] = newSeries;
  }
  return {
    seriesMap,
    datapointsMap,
    filteredUnits: (0, import_lodash_es.uniq)(filteredUnits),
    filteredTimeseries,
    originalFilteredUnits: (0, import_lodash_es.uniq)(originalFilteredUnits)
  };
}
