// @dynatrace/strato-components/charts/timeseries/hooks/useDataByYAxis.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useDataByYAxis_exports = {};
__export(useDataByYAxis_exports, {
  useDataByYAxis: () => useDataByYAxis
});
module.exports = __toCommonJS(useDataByYAxis_exports);
var import_lodash_es = require("lodash");
var import_useYAxisConfig = require("./useYAxisConfig.js");
function getAllUnits(data) {
  return (0, import_lodash_es.uniq)(
    data.shapePriority.flatMap((v) => data[v]?.map((s) => s.unit))
  );
}
function useDataByYAxis(data) {
  const yAxis = (0, import_useYAxisConfig.useYAxisConfig)();
  const allUnits = getAllUnits(data);
  const hasAnyAxisDefined = yAxis.left || yAxis.right;
  const hasOnlyOneAxisDefined = !yAxis.left || !yAxis.right;
  const shouldFilterData = allUnits.length > 1 && hasOnlyOneAxisDefined && hasAnyAxisDefined;
  if (shouldFilterData) {
    const [firstUnit] = allUnits;
    return {
      line: data.line.filter((series) => series.unit === firstUnit),
      bar: data.bar.filter((series) => series.unit === firstUnit),
      area: data.area.filter((series) => series.unit === firstUnit),
      band: data.band.filter((series) => series.unit === firstUnit),
      shapePriority: data.shapePriority
    };
  }
  return data;
}
