// @dynatrace/strato-components/charts/xy-chart/processing-strategies/utils/stacking/stack-datapoints.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var stack_datapoints_exports = {};
__export(stack_datapoints_exports, {
  getBaselineCoordinate: () => getBaselineCoordinate,
  getCombinedAxesOrientationKey: () => getCombinedAxesOrientationKey,
  getStackedValue: () => getStackedValue,
  getStackingData: () => getStackingData,
  isStackable: () => isStackable
});
module.exports = __toCommonJS(stack_datapoints_exports);
var import_guards = require("../../../types/guards.js");
var import_xy_chart = require("../../../types/xy-chart.js");
const orientationToBaselineAxis = {
  vertical: "x",
  horizontal: "y"
};
const getCombinedAxesOrientationKey = (xAxisId, yAxisId, chartOrientation) => `${xAxisId}-${yAxisId}-${chartOrientation}`;
const getStackingData = (chartOrientation, datapoint, xAxis, yAxis) => {
  const [start, end] = chartOrientation === "vertical" ? [datapoint.x0, datapoint.x1] : [datapoint.y0, datapoint.y1];
  const stackKey = (end ? `${start}-${end}` : `${start}`) + `-${getCombinedAxesOrientationKey(xAxis.id, yAxis.id, chartOrientation)}`;
  const value = chartOrientation === "vertical" ? datapoint.y1 : datapoint.x1;
  return { stackKey, value };
};
const getBaselineCoordinate = (orientation) => {
  if (!orientation || orientation === "both") {
    return;
  }
  return orientationToBaselineAxis[orientation];
};
const isStackable = (variant, chartOrientation, xAxis, yAxis) => {
  const baselineCoordinate = getBaselineCoordinate(chartOrientation);
  const isStackingAxisCategorical = baselineCoordinate && baselineCoordinate === "x" ? (0, import_guards.isCategoricalAxis)(yAxis) : (0, import_guards.isCategoricalAxis)(xAxis);
  return chartOrientation !== "both" && variant === import_xy_chart.BAR_VARIANT_STACK && !isStackingAxisCategorical;
};
const getStackedValue = (value, prevValue) => {
  return {
    minBoundary: value < 0 ? prevValue.minBoundary + value : prevValue.minBoundary,
    maxBoundary: value > 0 ? prevValue.maxBoundary + value : prevValue.maxBoundary
  };
};
