// @dynatrace/strato-components/charts/core/components/toolbar/ChartToolbarVisibilityOverlay.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ChartToolbarVisibilityOverlay_exports = {};
__export(ChartToolbarVisibilityOverlay_exports, {
  ChartToolbarVisibilityOverlay: () => ChartToolbarVisibilityOverlay
});
module.exports = __toCommonJS(ChartToolbarVisibilityOverlay_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_ChartToolbarRenderer_css = require("./ChartToolbarRenderer.sty.js");
var import_useApplyCallbackOnScroll = require("../../hooks/useApplyCallbackOnScroll.js");
var import_useEffectOnValueChange = require("../../hooks/useEffectOnValueChange.js");
var import_is_within_bounds = require("../../utils/is-within-bounds.js");
var import_InteractionsOverlay_css = require("../interactions-overlay/InteractionsOverlay.sty.js");
const ChartToolbarVisibilityOverlay = (props) => {
  const { children, visibleWhenMouseInBounds } = props;
  if (!visibleWhenMouseInBounds) {
    return children;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(VisibilityOverlay, { ...props });
};
const VisibilityOverlay = (props) => {
  const { boundSize, children, setIsVisible, isVisible, handlerRef } = props;
  const overlayRef = (0, import_react.useRef)(null);
  const overlayState = (0, import_react.useMemo)(() => {
    let bounds;
    return {
      inBounds: (e) => {
        if (!bounds) {
          bounds = overlayRef.current?.getBoundingClientRect();
        }
        return (0, import_is_within_bounds.isWithinBounds)(e, bounds);
      },
      resetBounds: () => {
        bounds = void 0;
      }
    };
  }, []);
  (0, import_useApplyCallbackOnScroll.useApplyCallbackOnScroll)(!isVisible, () => {
    overlayState.resetBounds();
  });
  (0, import_useEffectOnValueChange.useEffectOnValueChange)(() => {
    overlayState.resetBounds();
    setIsVisible(false);
  }, [boundSize.height, boundSize.width]);
  (0, import_react.useEffect)(() => {
    if (handlerRef?.current) {
      handlerRef.current.handleFocus = () => {
        setIsVisible(true);
      };
    }
    return () => {
      if (handlerRef?.current) {
        handlerRef.current.handleFocus = () => null;
      }
    };
  });
  (0, import_react.useEffect)(() => {
    const cbMouse = (event) => {
      if (!overlayState.inBounds(event)) {
        setIsVisible(false);
        overlayState.resetBounds();
      }
    };
    if (isVisible) {
      document.addEventListener("mouseup", cbMouse);
      document.addEventListener("pointerout", cbMouse);
    }
    return () => {
      document.removeEventListener("mouseup", cbMouse);
      document.removeEventListener("pointerout", cbMouse);
    };
  }, [isVisible]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      role: "none",
      ref: overlayRef,
      className: (0, import_clsx.default)(import_ChartToolbarRenderer_css.toolbarVisibilityOverlay, import_InteractionsOverlay_css.interactionsOverlay),
      onPointerMove: (e) => {
        if (overlayState.inBounds(e)) {
          isVisible || setIsVisible(true);
        }
      },
      onClick: () => {
        isVisible || setIsVisible(true);
      },
      children
    }
  );
};
