// @dynatrace/strato-components/charts/core/components/toolbar/ChartToolbarRenderer.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ChartToolbarRenderer_exports = {};
__export(ChartToolbarRenderer_exports, {
  ChartToolbarRenderer: () => ChartToolbarRenderer,
  InternalChartToolbarRenderer: () => InternalChartToolbarRenderer
});
module.exports = __toCommonJS(ChartToolbarRenderer_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_ChartToolbarRenderer_css = require("./ChartToolbarRenderer.sty.js");
var import_ChartToolbarVisibilityOverlay = require("./ChartToolbarVisibilityOverlay.js");
var import_CollapseButton = require("./CollapseButton.js");
var import_ChartDataMenu = require("./components/ChartDataMenu.js");
var import_Control = require("./components/Control.js");
var import_divider = require("./components/divider.js");
var import_IntentsCollapsedAndDownloadSection = require("./components/IntentsCollapsedAndDownloadSection.js");
var import_ToolbarMenu = require("./components/ToolbarMenu.js");
var import_constants = require("./constants.js");
var import_ChartToolbar_context = require("./context/ChartToolbar.context.js");
var import_useToolbar = require("./hooks/useToolbar.js");
var import_messages = require("./messages.js");
var import_does_fit_in_bounds = require("./utils/does-fit-in-bounds.js");
var import_estimate_toolbar_width = require("./utils/estimate-toolbar-width.js");
var import_get_toolbar_placement = require("./utils/get-toolbar-placement.js");
var import_Button = require("../../../../buttons/button/Button.js");
var import_useBehavioralTrackingProps = require("../../../../core/hooks/useBehavioralTrackingProps.js");
var import_useMergeRefs = require("../../../../core/hooks/useMergeRefs.js");
var import_Flex = require("../../../../layouts/flex/Flex.js");
var import_Tooltip = require("../../../../overlays/tooltip/Tooltip.js");
var import_A11yToolbarTabIndexContext = require("../../context/A11yToolbarTabIndexContext.js");
var import_useA11yToolbarControls = require("../../hooks/useA11yToolbarControls.js");
var import_useDragElement = require("../../hooks/useDragElement.js");
var import_useObserveContainerSize = require("../../hooks/useObserveContainerSize.js");
var import_get_draggable_bounds = require("../../utils/get-draggable-bounds.js");
var import_types = require("../interactions-overlay/types.js");
const InternalChartToolbarRenderer = (props) => {
  const [
    behaviorTrackingProps,
    {
      children,
      visible,
      placement,
      collapsed = false,
      draggable = true,
      removeDragHandle = false,
      initialA11yState,
      zIndex,
      onToggleToolbar,
      onToggleMenu,
      onMouseEnter,
      onMouseLeave,
      menuIcon,
      menuTriggerLabel,
      menuDisplayMode = "collapsed",
      onClose,
      isDownloadEnabled = false,
      intents,
      handleDownloadDataToolbar,
      toolbarDisplayMode = "overlay",
      visibleWhenMouseInBounds = false,
      overlayClassName = import_ChartToolbarRenderer_css.defaultOverlayPadding,
      intentsAsyncTemplate,
      handlerRef
    }
  ] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(props);
  const {
    isExpanded,
    lastItemSelected,
    updateLastItemSelected,
    toggleExpanded,
    isCollapsable,
    fitsBounds,
    updateFitsBounds,
    shouldShowMenu
  } = (0, import_useToolbar.useToolbar)(collapsed, menuDisplayMode);
  const isInlineMode = toolbarDisplayMode === "inline";
  const [isMenuOpen, setIsMenuOpen] = (0, import_react.useState)(void 0);
  const [toolbarElement, setToolbarElement] = (0, import_react.useState)();
  const [boundSize, boundRef] = (0, import_useObserveContainerSize.useObserveContainerSize)();
  const [toolbarSize, toolbarRefCallback] = (0, import_useObserveContainerSize.useObserveContainerSize)();
  const intl = (0, import_react_intl.useIntl)();
  const mouseInBounds = (0, import_react.useRef)(false);
  const [isVisible, setIsVisible] = (0, import_react.useState)(false);
  const toolbarElementRef = (0, import_react.useCallback)((toolbarElement2) => {
    setToolbarElement(toolbarElement2);
  }, []);
  const toolbarRef = (0, import_useMergeRefs.useMergeRefs)([toolbarRefCallback, toolbarElementRef]);
  const { a11yControlsValue, setKeyEvent, setActiveControl } = (0, import_useA11yToolbarControls.useA11yToolbarControls)(initialA11yState);
  const maxBounds = (0, import_react.useMemo)(
    () => (0, import_get_draggable_bounds.getDraggableBounds)(toolbarSize, boundSize),
    [toolbarSize, boundSize]
  );
  (0, import_react.useEffect)(() => {
    updateFitsBounds(
      (0, import_does_fit_in_bounds.doesFitInBounds)((0, import_estimate_toolbar_width.estimateToolbarWidth)(children), boundSize)
    );
  }, [children, boundSize]);
  (0, import_react.useEffect)(() => {
    toggleExpanded(fitsBounds && !collapsed);
  }, [collapsed, fitsBounds]);
  const initialPosition = (0, import_get_toolbar_placement.getToolbarPlacement)(placement, maxBounds);
  const { isDragging, triggerDragRef, didDrag, setDidDrag } = (0, import_useDragElement.useDragElement)(
    toolbarElement,
    maxBounds,
    initialPosition
  );
  const expandHandler = (0, import_react.useCallback)(
    (isExpanded2) => {
      toggleExpanded(isExpanded2);
      onToggleToolbar?.(isExpanded2);
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [onToggleToolbar]
  );
  const handleToggleMenu = (open) => {
    if (!open) {
      if (!mouseInBounds.current) {
        onMouseLeave?.();
      } else {
        onMouseEnter?.();
      }
    }
    setIsMenuOpen(open);
    onToggleMenu?.(open);
  };
  const collapseButton = (0, import_react.useMemo)(
    () => Array.isArray(children) && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_CollapseButton.CollapseButton,
      {
        onClick: (isExpanded2) => {
          expandHandler(isExpanded2);
          setActiveControl("explore");
        }
      }
    ),
    [children, expandHandler, setActiveControl]
  );
  const expandButton = (0, import_react.useMemo)(
    () => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Control.Control,
      {
        prefixIcon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ChevronRightDoubleIcon, {}),
        tooltipText: intl.formatMessage(import_messages.TOOLBAR_MESSAGES.chartToolbarExpand),
        onClick: () => expandHandler(true),
        tabIndex: -1,
        testId: "toolbar-expand-button",
        "data-dt-name": "ToolbarExpand"
      }
    ),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [expandHandler]
  );
  const onMouseEnterHandler = () => {
    mouseInBounds.current = true;
    onMouseEnter?.();
  };
  const onMouseLeaveHandler = (event) => {
    event.stopPropagation();
    mouseInBounds.current = false;
    onMouseLeave?.();
  };
  const handleOnMenuEnter = () => {
    mouseInBounds.current = false;
  };
  const handleOnMenuLeave = () => {
    mouseInBounds.current = true;
  };
  const onCloseAutoFocus = (e) => {
    if ((lastItemSelected === import_types.INTERACTION_MODE.Explore && onClose || lastItemSelected === import_types.INTERACTION_MODE.Zoom) && onClose) {
      e.preventDefault();
      onClose();
      updateLastItemSelected(void 0);
    }
  };
  const handleResetDidDrag = (0, import_react.useCallback)(() => {
    setDidDrag(false);
  }, [setDidDrag]);
  const showCollapsedIntentAndDownloadSubmenu = (!isExpanded || !fitsBounds) && (!!intents?.length || isDownloadEnabled);
  const showExpandButton = !isExpanded && fitsBounds;
  const showCollapseButton = isExpanded && isCollapsable;
  const showIntentMenu = isExpanded && fitsBounds && !!(intents?.length || intentsAsyncTemplate);
  const toolbarMenuConfig = (0, import_react.useMemo)(
    () => ({
      isMenuElement: true,
      isExpanded,
      lastItemSelected,
      updateLastItemSelected
    }),
    [isExpanded, lastItemSelected, updateLastItemSelected]
  );
  const intentsMenuConfig = (0, import_react.useMemo)(
    () => ({
      isMenuElement: false,
      isExpanded,
      lastItemSelected,
      didDrag,
      resetDidDrag: handleResetDidDrag,
      updateLastItemSelected
    }),
    [
      isExpanded,
      lastItemSelected,
      didDrag,
      handleResetDidDrag,
      updateLastItemSelected
    ]
  );
  const handleKeyDown = (event) => {
    if (event.target === triggerDragRef.current && event.code !== "Tab") {
      event.stopPropagation();
      event.preventDefault();
      return;
    }
    if (event.code === "ArrowLeft" || event.code === "ArrowRight") {
      event.stopPropagation();
      setKeyEvent(event);
    }
    if (event.code === "Enter") {
      event.stopPropagation();
    }
  };
  const handleOnBlur = (event) => {
    event.stopPropagation();
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ChartToolbarVisibilityOverlay.ChartToolbarVisibilityOverlay,
    {
      handlerRef,
      boundSize,
      setIsVisible,
      visibleWhenMouseInBounds,
      isVisible,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        "div",
        {
          className: (0, import_clsx.default)(
            (0, import_ChartToolbarRenderer_css.chartToolbarArea)({
              inline: isInlineMode
            }),
            overlayClassName
          ),
          "data-testid": "chart-plot-overlay-toolbar",
          ref: boundRef,
          style: { zIndex },
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "div",
            {
              className: (0, import_ChartToolbarRenderer_css.chartToolbar)({ inline: isInlineMode }),
              id: "chart-toolbar",
              ref: toolbarRef,
              role: "toolbar",
              "data-testid": "chart-toolbar-container",
              onBlur: handleOnBlur,
              onKeyDown: handleKeyDown,
              onMouseEnter: onMouseEnterHandler,
              onMouseLeave: onMouseLeaveHandler,
              ...behaviorTrackingProps,
              children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_A11yToolbarTabIndexContext.A11yToolbarTabIndexContext.Provider, { value: a11yControlsValue, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                "div",
                {
                  className: (0, import_ChartToolbarRenderer_css.outerContainer)({
                    visible: visible || isMenuOpen || isDragging || visibleWhenMouseInBounds && isVisible,
                    inline: isInlineMode
                  }),
                  "data-testid": "chart-toolbar-wrapper",
                  children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { gap: 0, flexDirection: "row", alignItems: "center", children: [
                    draggable && !removeDragHandle && [
                      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                        import_Tooltip.Tooltip,
                        {
                          text: intl.formatMessage(
                            import_messages.TOOLBAR_MESSAGES.chartToolbarMove
                          ),
                          disabled: isDragging,
                          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                            import_Button.Button,
                            {
                              className: import_ChartToolbarRenderer_css.grip,
                              ref: triggerDragRef,
                              variant: "default",
                              size: "condensed",
                              "data-testid": "chart-toolbar-grip",
                              "aria-label": import_constants.TOOLBAR_DRAG_BUTTON_ARIA_LABEL,
                              children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.DragAllDirectionIcon, {}) })
                            }
                          )
                        },
                        "tooltip"
                      ),
                      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_divider.VerticalDivider, {}, "divider")
                    ],
                    /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
                      "div",
                      {
                        "data-testid": "toolbar-children",
                        style: { display: "flex", alignItems: "center" },
                        ref: draggable && removeDragHandle ? triggerDragRef : void 0,
                        children: [
                          shouldShowMenu && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ChartToolbar_context.ChartToolbarContext.Provider, { value: toolbarMenuConfig, children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
                            import_ToolbarMenu.ToolbarMenu,
                            {
                              onMenuOpen: handleToggleMenu,
                              activeModeIcon: menuIcon,
                              triggerLabel: menuTriggerLabel,
                              onMenuEnter: handleOnMenuEnter,
                              onMenuLeave: handleOnMenuLeave,
                              menuAlignment: isExpanded ? "start" : "end",
                              onCloseAutoFocus,
                              "data-dt-name": "ToolbarTools",
                              "data-dt-properties": isExpanded && fitsBounds ? "state:changeMode" : "state:tools",
                              children: [
                                children,
                                /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                                  import_IntentsCollapsedAndDownloadSection.IntentsCollapsedAndDownloadSection,
                                  {
                                    hidden: !showCollapsedIntentAndDownloadSubmenu,
                                    intents,
                                    showDownloadButton: isDownloadEnabled,
                                    handleDownloadDataToolbar
                                  }
                                ),
                                showExpandButton && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
                                  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_divider.HorizontalDivider, {}),
                                  expandButton
                                ] })
                              ]
                            }
                          ) }),
                          /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_ChartToolbar_context.ChartToolbarContext.Provider, { value: intentsMenuConfig, children: [
                            children,
                            showIntentMenu && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                              import_ChartDataMenu.ChartDataMenu,
                              {
                                onMenuOpen: handleToggleMenu,
                                onMenuEnter: handleOnMenuEnter,
                                onMenuLeave: handleOnMenuLeave,
                                isDownloadEnabled,
                                handleDownloadDataToolbar,
                                intents,
                                showDivider: !!menuIcon,
                                isDragging,
                                buttonMode: isInlineMode ? "inline" : "standard",
                                intentsAsyncTemplate
                              }
                            )
                          ] }),
                          showCollapseButton && collapseButton
                        ]
                      }
                    )
                  ] })
                }
              ) })
            }
          )
        }
      )
    }
  );
};
InternalChartToolbarRenderer["displayName"] = "ChartToolbarRenderer";
const ChartToolbarRenderer = Object.assign(
  InternalChartToolbarRenderer,
  {
    Control: import_Control.Control
  }
);
