// @dynatrace/strato-components/charts/core/components/toolbar/CollapseButton.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CollapseButton_exports = {};
__export(CollapseButton_exports, {
  CollapseButton: () => CollapseButton
});
module.exports = __toCommonJS(CollapseButton_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_Control = require("./components/Control.js");
var import_messages = require("./messages.js");
var import_useA11yToolbarTabIndex = require("../../hooks/useA11yToolbarTabIndex.js");
const CollapseButton = ({ onClick }) => {
  const buttonRef = (0, import_react.useRef)(null);
  const intl = (0, import_react_intl.useIntl)();
  const tabIndex = (0, import_useA11yToolbarTabIndex.useA11yToolbarTabIndex)("collapse");
  const onClickHandler = () => {
    onClick(false);
  };
  const onKeyDownHandler = (event) => {
    if (event.key === "Enter") {
      onClick(false);
    }
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Control.Control,
    {
      ref: buttonRef,
      testId: "toolbar-collapse-button",
      prefixIcon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ChevronLeftDoubleIcon, {}),
      tooltipText: intl.formatMessage(import_messages.TOOLBAR_MESSAGES.chartToolbarCollapse),
      onKeyDown: onKeyDownHandler,
      onClick: onClickHandler,
      tabIndex,
      divider: "left",
      "data-dt-name": "ToolbarCollapse"
    }
  );
};
CollapseButton["displayName"] = "CollapseButton";
