// @dynatrace/strato-components/charts/core/components/toolbar/messages.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var messages_exports = {};
__export(messages_exports, {
  TOOLBAR_MESSAGES: () => TOOLBAR_MESSAGES
});
module.exports = __toCommonJS(messages_exports);
var import_react_intl = require("react-intl");
const TOOLBAR_MESSAGES = (0, import_react_intl.defineMessages)({
  chartToolbarMove: {
    id: "Rx7WtBPx/tyQ5h+j",
    defaultMessage: "Move",
    description: 'Text in a tooltip shown when hovering "Move" action in the time series chart toolbar'
  },
  chartToolbarCollapse: {
    id: "dTJ8Pdj8Xm9Uh9Y+",
    defaultMessage: "Collapse toolbar",
    description: 'Text in a tooltip shown when hovering "Collapse" action in the time series chart toolbar'
  },
  chartToolbarExpand: {
    id: "Y0Ru/2gdHx6uuwvl",
    defaultMessage: "Expand toolbar",
    description: 'Text in a tooltip shown when hovering "Expand" action in the time series chart toolbar'
  },
  downloadDataButton: {
    id: "f4/NcUWxDYnAQ68s",
    defaultMessage: "Download as CSV",
    description: 'Text in "Download data" button in the chart toolbar'
  },
  downloadData: {
    id: "u5xmuzxNDovVbbQX",
    defaultMessage: "Download as CSV",
    description: 'Text in a tooltip shown when hovering the "Download data" button in the time series chart toolbar'
  },
  moreOptions: {
    id: "eONO7S5ZXY7V+R6d",
    defaultMessage: "More options",
    description: "Text in button that opens a menu with more actions in the chart toolbar"
  },
  changeMode: {
    id: "uDo0yY4voNQ3eiN7",
    defaultMessage: "Change mode",
    description: "Text in a tooltip shown when hovering the expanded toolbar's menu button in the time series chart toolbar"
  },
  chartTools: {
    id: "LgOM6mSxWTUB+NI/",
    defaultMessage: "Chart tools",
    description: "Text in a tooltip shown when hovering the manually collapsed toolbar's menu button in the time series chart toolbar"
  },
  or: {
    id: "S3llhud3ydpS6ObI",
    defaultMessage: "or",
    description: "or"
  }
});
