// @dynatrace/strato-components/esm/charts/categorical-bar/components/axes/CategoricalBarChartCategoryAxis.js
import { jsx } from "react/jsx-runtime";
import { useMemo } from "react";
import { getTickCategories } from "./utils/categorical-category-axis.js";
import { CategoryBottomAxis } from "../../../core/components/axes/components/CategoryBottomAxis.js";
import { CategoryLeftAxis } from "../../../core/components/axes/components/CategoryLeftAxis.js";
import { ChartGrid } from "../../../core/components/chart-grid/ChartGridRoot.js";
import { useCategoricalBarChartBaseConfig } from "../../hooks/useCategoricalBarChartBaseConfig.js";
import { useCategoricalBarChartCategoryAxisConfig } from "../../hooks/useCategoricalBarChartCategoryAxisConfig.js";
import { useCategoricalBarChartScales } from "../../hooks/useCategoricalBarChartScales.js";
import { useCategoricalBarFilteredData } from "../../hooks/useCategoricalBarFilteredData.js";
import { useTruncationMode } from "../../hooks/useTruncationMode.js";
import { AREAS } from "../../types/categorical-bar-area-rules.js";
import { isVertical } from "../../utils/layout.js";
const CategoricalBarChartCategoryAxis = () => {
  const { filteredCategories, categoryMap } = useCategoricalBarFilteredData();
  const { categoryScale } = useCategoricalBarChartScales();
  const { layout } = useCategoricalBarChartBaseConfig();
  const truncationMode = useTruncationMode();
  const { label: categoryAxisLabel, tickLabelLayout } = useCategoricalBarChartCategoryAxisConfig();
  const categoryTicks = useMemo(
    () => getTickCategories(filteredCategories, categoryMap),
    [categoryMap, filteredCategories]
  );
  return isVertical(layout) ? /* @__PURE__ */ jsx(ChartGrid.GridArea, { gridArea: AREAS.bottomAxis, children: ({ availableWidth, availableHeight }) => /* @__PURE__ */ jsx(
    CategoryBottomAxis,
    {
      isVerticalLayout: true,
      categories: categoryTicks,
      containerHeight: availableHeight,
      containerWidth: availableWidth,
      scale: categoryScale,
      label: categoryAxisLabel,
      tickLabelLayout,
      truncationMode
    }
  ) }) : /* @__PURE__ */ jsx(ChartGrid.GridArea, { gridArea: AREAS.leftAxis, children: ({ availableWidth, availableHeight }) => /* @__PURE__ */ jsx(
    CategoryLeftAxis,
    {
      isVerticalLayout: false,
      categories: categoryTicks,
      containerHeight: availableHeight,
      containerWidth: availableWidth,
      scale: categoryScale,
      label: categoryAxisLabel,
      truncationMode
    }
  ) });
};
CategoricalBarChartCategoryAxis["displayName"] = "CategoricalBarChartAxis";
export {
  CategoricalBarChartCategoryAxis
};
