// @dynatrace/strato-components/esm/charts/categorical-bar/components/axes/CategoricalBarChartValueAxis.js
import { jsx } from "react/jsx-runtime";
import { ValueBottomAxis } from "../../../core/components/axes/components/ValueBottomAxis.js";
import { ValueLeftAxis } from "../../../core/components/axes/components/ValueLeftAxis.js";
import { ChartGrid } from "../../../core/components/chart-grid/ChartGridRoot.js";
import { useCategoricalBarChartBaseConfig } from "../../hooks/useCategoricalBarChartBaseConfig.js";
import { useCategoricalBarChartScales } from "../../hooks/useCategoricalBarChartScales.js";
import { useCategoricalBarChartTicks } from "../../hooks/useCategoricalBarChartTicks.js";
import { useCategoricalBarChartValueAxisConfig } from "../../hooks/useCategoricalBarChartValueAxisConfig.js";
import { AREAS } from "../../types/categorical-bar-area-rules.js";
import { isVertical } from "../../utils/layout.js";
const CategoricalBarChartValueAxis = () => {
  const { valueScale } = useCategoricalBarChartScales();
  const { label: valueAxisLabel } = useCategoricalBarChartValueAxisConfig();
  const { layout } = useCategoricalBarChartBaseConfig();
  const { valueAxis } = useCategoricalBarChartTicks();
  return isVertical(layout) ? /* @__PURE__ */ jsx(ChartGrid.GridArea, { gridArea: AREAS.leftAxis, children: ({ availableWidth, availableHeight }) => /* @__PURE__ */ jsx(
    ValueLeftAxis,
    {
      formattedTicks: valueAxis.formattedTicks,
      ticks: valueAxis.ticks,
      containerHeight: availableHeight,
      containerWidth: availableWidth,
      scale: valueScale,
      label: valueAxisLabel
    }
  ) }) : /* @__PURE__ */ jsx(ChartGrid.GridArea, { gridArea: AREAS.bottomAxis, children: ({ availableWidth, availableHeight }) => /* @__PURE__ */ jsx(
    ValueBottomAxis,
    {
      formattedTicks: valueAxis.formattedTicks,
      ticks: valueAxis.ticks,
      containerHeight: availableHeight,
      containerWidth: availableWidth,
      scale: valueScale,
      label: valueAxisLabel
    }
  ) });
};
CategoricalBarChartValueAxis["displayName"] = "CategoricalBarChartValueAxis";
export {
  CategoricalBarChartValueAxis
};
