// @dynatrace/strato-components/esm/charts/categorical-bar/components/interactions/utils/data-points.js
import { OVERLAY_PADDING } from "../../../../core/components/data-point/constants.js";
import { getGroupedDataPointsByCoordinate } from "../../../../core/utils/group-by-threshold.js";
import { isVertical } from "../../../utils/layout.js";
const mapToDataPoint = (dimension) => ({
  ...dimension,
  x: dimension.x + OVERLAY_PADDING,
  y: dimension.y + OVERLAY_PADDING,
  color: dimension.datapoint.color,
  id: dimension.datapoint.id
});
const filterOutOfChartBounds = (chartLimit, coordinate) => (dimension) => dimension[coordinate] >= OVERLAY_PADDING && dimension[coordinate] <= chartLimit + OVERLAY_PADDING;
function getGroupDataPointsBy(groupDataPoints, dimensionId) {
  return groupDataPoints?.filter(
    (points) => points.some((d) => d.id === dimensionId)
  );
}
function generateGroupDataPoints(points, isGroupAllowed, layout, chartLimit) {
  const coordinate = isVertical(layout) ? "y" : "x";
  const mappedPoints = points.map(mapToDataPoint).filter(filterOutOfChartBounds(chartLimit, coordinate));
  if (!isGroupAllowed) {
    return mappedPoints.map((point) => [point]);
  }
  return getGroupedDataPointsByCoordinate(mappedPoints, coordinate);
}
export {
  filterOutOfChartBounds,
  generateGroupDataPoints,
  getGroupDataPointsBy,
  mapToDataPoint
};
