// @dynatrace/strato-components/charts/meter-bar/shared/utils/slot-handlers.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var slot_handlers_exports = {};
__export(slot_handlers_exports, {
  handleColorRuleSlot: () => handleColorRuleSlot,
  handleIntentSlot: () => handleIntentSlot,
  handleLabelSlot: () => handleLabelSlot,
  handleMaxSlot: () => handleMaxSlot,
  handleMinSlot: () => handleMinSlot,
  handleToolbarSlot: () => handleToolbarSlot,
  handleValueSlot: () => handleValueSlot
});
module.exports = __toCommonJS(slot_handlers_exports);
var import_lodash_es = require("lodash");
var import_ColorRule = require("../../../core/slots/ColorRule.js");
var import_extract_behavior_tracking_props = require("../../../core/utils/extract-behavior-tracking-props.js");
function handleLabelSlot(child, slots) {
  slots.label = child.props.children;
}
function handleValueSlot(child, slots) {
  slots.value = {
    value: child.props.children,
    auto: (0, import_lodash_es.isUndefined)(child.props.children)
  };
}
function handleMinSlot(child, slots) {
  slots.min = {
    value: child.props.children,
    auto: (0, import_lodash_es.isUndefined)(child.props.children)
  };
}
function handleMaxSlot(child, slots) {
  slots.max = {
    value: child.props.children,
    auto: (0, import_lodash_es.isUndefined)(child.props.children)
  };
}
function handleColorRuleSlot(child, slots) {
  const colorRule = child.props;
  if ((0, import_ColorRule.isValidColorRule)(colorRule)) {
    slots.colorRules.push(colorRule);
  }
}
function handleIntentSlot(child, intents) {
  return intents ? [...intents, child.props] : [child.props];
}
function handleToolbarSlot(child, slots) {
  if (slots.toolbar && "hidden" in child.props) {
    slots.toolbar.hidden = child.props.hidden;
  }
  if (slots.toolbar) {
    Object.assign(slots.toolbar, (0, import_extract_behavior_tracking_props.extractBehaviorTrackingProps)(child.props));
  }
}
