// @dynatrace/strato-components/filters/filter-field/state/recent-pinned-store.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var recent_pinned_store_exports = {};
__export(recent_pinned_store_exports, {
  addFilterToState: () => addFilterToState,
  createRecentAndPinnedFilterStore: () => createRecentAndPinnedFilterStore,
  removeFilterFromState: () => removeFilterFromState,
  scopedRecentAndPinnedFiltersStores: () => scopedRecentAndPinnedFiltersStores,
  voidStore: () => voidStore
});
module.exports = __toCommonJS(recent_pinned_store_exports);
var import_zustand = require("zustand");
var import_tree_conversion = require("../utils/conversion/tree-conversion.js");
function addFilterToState(storedFilters, filter, limit) {
  const trimmedFilter = filter.trim();
  const foundIndex = storedFilters.findIndex(
    (saved) => saved.filter === trimmedFilter
  );
  const newValues = [...storedFilters];
  if (foundIndex === -1) {
    newValues.push({
      filter: trimmedFilter,
      parserVersion: import_tree_conversion.parserVersion,
      timeUsed: /* @__PURE__ */ new Date()
    });
  } else if (foundIndex > -1) {
    if (storedFilters[foundIndex].parserVersion !== import_tree_conversion.parserVersion) {
      newValues[foundIndex].parserVersion = import_tree_conversion.parserVersion;
    }
    storedFilters[foundIndex].timeUsed = /* @__PURE__ */ new Date();
  }
  newValues.sort((a, b) => b.timeUsed.getTime() - a.timeUsed.getTime());
  return newValues.slice(0, limit);
}
function removeFilterFromState(storedFilters, filter) {
  const trimmedFilter = filter.trim();
  const foundIndex = storedFilters.findIndex(
    (saved) => saved.filter === trimmedFilter
  );
  const newValues = [...storedFilters];
  if (foundIndex > -1) {
    newValues.splice(foundIndex, 1);
  }
  return newValues;
}
function createRecentAndPinnedFilterStore() {
  return (0, import_zustand.createStore)((set, get) => ({
    canPin: true,
    setCanPin: (canPin) => {
      set(() => ({ canPin }));
    },
    recentFilters: [],
    setRecentFilters: (filters) => {
      set(() => ({ recentFilters: filters }));
    },
    saveRecentFilter: (filterString) => {
      set((state) => ({
        recentFilters: addFilterToState(state.recentFilters, filterString, 100)
      }));
      return get().recentFilters;
    },
    pinnedFilters: [],
    setPinnedFilters: (filters) => {
      set(() => ({ pinnedFilters: filters }));
    },
    pinFilter: (filterString) => {
      set((state) => ({
        pinnedFilters: addFilterToState(state.pinnedFilters, filterString, 10)
      }));
      return get().pinnedFilters;
    },
    unpinFilter: (filterString) => {
      set((state) => ({
        pinnedFilters: removeFilterFromState(state.pinnedFilters, filterString)
      }));
      return get().pinnedFilters;
    }
  }));
}
const voidStore = (0, import_zustand.createStore)(() => ({
  canPin: true,
  setCanPin: () => void 0,
  recentFilters: [],
  setRecentFilters: () => void 0,
  saveRecentFilter: (filter) => [],
  pinnedFilters: [],
  setPinnedFilters: (filters) => void 0,
  pinFilter: (filter) => [],
  unpinFilter: (filter) => []
}));
const scopedRecentAndPinnedFiltersStores = /* @__PURE__ */ new Map([["root", createRecentAndPinnedFilterStore()]]);
