// @dynatrace/strato-components/charts/annotations-chart/AnnotationsChart.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var AnnotationsChart_exports = {};
__export(AnnotationsChart_exports, {
  AnnotationsChart: () => AnnotationsChart,
  _AnnotationsChart: () => _AnnotationsChart
});
module.exports = __toCommonJS(AnnotationsChart_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_AnnotationsDisplay = require("./components/AnnotationsDisplay.js");
var import_AnnotationsAxis = require("./components/axis/AnnotationsAxis.js");
var import_constants = require("./constants.js");
var import_useAnnotationsChartConfig = require("./hooks/useAnnotationsChartConfig.js");
var import_annotations_chart_config_provider = require("./providers/annotations-chart-config.provider.js");
var import_Marker = require("./slots/Marker.js");
var import_Tooltip = require("./slots/Tooltip.js");
var import_Track = require("./slots/Track.js");
var import_XAxis = require("./slots/XAxis.js");
var import_iterate_slots = require("./utils/iterate-slots.js");
var import_scales = require("./utils/scales.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_defaultChartValues_css = require("../../core/styles/defaultChartValues.sty.js");
var import_AnnotationsScrollWrapper = require("../core/components/annotations/components/overlays/AnnotationsScrollWrapper.js");
var import_constants2 = require("../core/components/annotations/constants.js");
var import_is_numerical_marker = require("../core/components/annotations/utils/is-numerical-marker.js");
var import_ChartStateHandler = require("../core/components/chart-states/ChartStateHandler.js");
var import_get_chart_states_template = require("../core/components/chart-states/utils/get-chart-states-template.js");
var import_DataDtrumContext = require("../core/context/DataDtrumContext.js");
var import_useObserveContainerSize = require("../core/hooks/useObserveContainerSize.js");
var import_EmptyStateSlot = require("../core/slots/chart-states/EmptyStateSlot.js");
var import_ErrorStateSlot = require("../core/slots/chart-states/ErrorStateSlot.js");
var import_chart_types = require("../core/types/chart-types.js");
var import_coerceSizeValue = require("../core/utils/coerceSizeValue.js");
const AXIS_HEIGHT = 25;
const X_AXIS_LABEL_HEIGHT = 10;
const COMPONENT_NAME = "AnnotationsChart";
const calculateXAxisHeight = (isHidden, label) => isHidden ? 0 : label ? AXIS_HEIGHT + X_AXIS_LABEL_HEIGHT : AXIS_HEIGHT;
const _AnnotationsChart = (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    annotationsActions,
    children,
    loading = false,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "data-testid": dataTestId,
    style: consumerStyle,
    className: consumerClassName,
    width: propsWidth
  } = props;
  const [behaviorTrackingProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(props);
  const [containerDimension, sizeObserverRef] = (0, import_useObserveContainerSize.useObserveContainerSize)();
  const { aggregatedMarkers, tracks, xAxisProps, tooltipProps } = (0, import_iterate_slots.iterateSlots)(children);
  const { emptyState, errorState } = (0, import_get_chart_states_template.getChartStatesTemplates)(children);
  const hasNumericalData = aggregatedMarkers.length > 0 && (0, import_is_numerical_marker.isNumericalData)(aggregatedMarkers);
  const config = (0, import_useAnnotationsChartConfig.useAnnotationsChartConfig)(
    props,
    xAxisProps,
    tooltipProps,
    hasNumericalData
  );
  const {
    showLabels,
    xAxisConfig,
    tooltipConfig,
    textOverflow,
    truncateMode,
    maxLines
  } = config;
  const shouldShowLabels = tracks.some(({ label }) => label) && showLabels;
  const labelsOffset = shouldShowLabels ? import_constants2.DEFAULT_LABELS_OFFSET + config.labelWidth : 0;
  const scale = (0, import_scales.buildScale)(
    aggregatedMarkers,
    containerDimension.width - labelsOffset,
    xAxisConfig
  );
  const shownTracks = tracks.filter(({ hidden }) => !hidden);
  const tracksHeight = shownTracks.length * import_constants2.TRACK_HEIGHT + import_constants2.TRACK_SEPARATOR_STROKE_WIDTH;
  const axisHeight = calculateXAxisHeight(
    xAxisConfig.hidden,
    config.xAxisConfig.label
  );
  const coercedHeight = (0, import_coerceSizeValue.coerceSizeValue)(config.height);
  const coercedWidth = (0, import_coerceSizeValue.coerceSizeValue)(propsWidth);
  const height = coercedHeight ?? tracksHeight + axisHeight;
  const width = coercedWidth ?? "100%";
  const scrollWrapperHeight = Math.min(
    tracksHeight,
    Math.max(0, containerDimension.height - axisHeight)
  );
  const containerRef = (0, import_react.useRef)(null);
  const rootRef = (0, import_useMergeRefs.useMergeRefs)([
    forwardedRef,
    containerRef,
    sizeObserverRef
  ]);
  (0, import_react.useImperativeHandle)(forwardedRef, () => ({
    element: containerRef.current,
    getConfig: () => {
      return JSON.stringify({
        type: import_chart_types.ChartType.AnnotationsChart,
        ...config
      });
    }
  }));
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ChartStateHandler.ChartStateHandler,
    {
      empty: (0, import_lodash_es.isEmpty)(tracks),
      height: "100%",
      emptyState,
      errorState,
      isLoading: loading,
      className: (0, import_clsx.default)(import_defaultChartValues_css.defaultChartSize, consumerClassName),
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "data-testid": dataTestId,
      "data-dt-component": COMPONENT_NAME,
      style: consumerStyle,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        "div",
        {
          style: { height, width },
          ref: rootRef,
          "aria-label": import_constants.ANNOTATIONS_CHART_ARIA_LABEL,
          role: "img",
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_annotations_chart_config_provider.AnnotationsChartConfigProvider, { config, children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_DataDtrumContext.DataDtrumContext.Provider, { value: { dataDtrumAllow, dataDtrumMask }, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_AnnotationsScrollWrapper.AnnotationsScrollWrapper,
              {
                height: scrollWrapperHeight,
                width: containerDimension.width,
                children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  import_AnnotationsDisplay.AnnotationsDisplay,
                  {
                    tracks,
                    scale,
                    annotationsActions,
                    showLabels: shouldShowLabels,
                    labelsOffset,
                    width: containerDimension.width,
                    tooltipConfig,
                    textOverflow,
                    truncateMode,
                    maxLines,
                    ...behaviorTrackingProps
                  }
                )
              }
            ) }),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_AnnotationsAxis.AnnotationsAxis,
              {
                height: axisHeight,
                scale,
                width: containerDimension.width,
                labelsOffset
              }
            )
          ] })
        }
      )
    }
  );
});
const AnnotationsChart = Object.assign(_AnnotationsChart, {
  Track: import_Track.Track,
  Marker: import_Marker.Marker,
  XAxis: import_XAxis.XAxis,
  Tooltip: import_Tooltip.Tooltip,
  EmptyState: import_EmptyStateSlot.EmptyStateSlot,
  ErrorState: import_ErrorStateSlot.ErrorStateSlot
});
