// @dynatrace/strato-components/charts/annotations-chart/constants.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var constants_exports = {};
__export(constants_exports, {
  ANNOTATIONS_CHART_ARIA_LABEL: () => ANNOTATIONS_CHART_ARIA_LABEL,
  DEFAULT_ANNOTATIONS_CHART_CONFIG: () => DEFAULT_ANNOTATIONS_CHART_CONFIG,
  DEFAULT_AXIS_HIDDEN: () => DEFAULT_AXIS_HIDDEN,
  DEFAULT_NUMERICAL_AXIS_TYPE: () => DEFAULT_NUMERICAL_AXIS_TYPE,
  DEFAULT_TOOLTIP_HIDDEN: () => DEFAULT_TOOLTIP_HIDDEN
});
module.exports = __toCommonJS(constants_exports);
var import_constants = require("../core/components/annotations/constants.js");
const DEFAULT_AXIS_HIDDEN = false;
const DEFAULT_NUMERICAL_AXIS_TYPE = "linear";
const DEFAULT_TOOLTIP_HIDDEN = false;
const DEFAULT_ANNOTATIONS_CHART_CONFIG = {
  showLabels: import_constants.DEFAULT_SHOW_LABELS,
  height: void 0,
  labelWidth: import_constants.DEFAULT_LABELS_WIDTH,
  xAxisConfig: {
    hidden: DEFAULT_AXIS_HIDDEN
  },
  tooltipConfig: {
    hidden: DEFAULT_TOOLTIP_HIDDEN,
    template: void 0
  },
  textOverflow: import_constants.DEFAULT_TEXT_OVERFLOW,
  truncateMode: import_constants.DEFAULT_TRUNCATION_MODE,
  maxLines: import_constants.DEFAULT_TOOLTIP_MAX_LINES
};
const ANNOTATIONS_CHART_ARIA_LABEL = "Annotations Chart";
