// @dynatrace/strato-components/charts/core/components/legend/categorical/components/ChartLegendSeriesActions/ChartLegendSeriesActions.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ChartLegendSeriesActions_exports = {};
__export(ChartLegendSeriesActions_exports, {
  ChartLegendSeriesActions: () => ChartLegendSeriesActions
});
module.exports = __toCommonJS(ChartLegendSeriesActions_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_ChartLegendSeriesActions_css = require("./ChartLegendSeriesActions.sty.js");
var import_Button = require("../../../../../../../buttons/button/Button.js");
var import_Menu = require("../../../../../../../navigation/menu/Menu.js");
var import_useCustomSeriesActions = require("../../../../../hooks/useCustomSeriesActions.js");
var import_render_menu_action_children = require("../../../../../utils/render-menu-action-children.js");
const ChartLegendSeriesActions = (props) => {
  const {
    disabled,
    onOpenChange,
    isOpen,
    visible = false,
    onFocus,
    onBlur,
    onMouseEnter,
    onMouseLeave,
    defaultActions,
    actions
  } = props;
  const handleFocus = (event) => {
    event.stopPropagation();
    onFocus();
  };
  const customActions = (0, import_useCustomSeriesActions.useCustomSeriesActions)(actions, isOpen);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Menu.Menu, { onOpenChange, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Trigger, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Button.Button,
      {
        "data-testid": "legend-more-actions-button",
        variant: "default",
        disabled,
        onFocus: handleFocus,
        onBlur,
        onMouseEnter,
        onMouseLeave,
        className: (0, import_ChartLegendSeriesActions_css.seriesButton)({ visible }),
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.DotMenuIcon, { size: "small" }) })
      }
    ) }),
    !disabled && isOpen && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Menu.Menu.Content, { "data-testid": "menu-content", children: [
      defaultActions ? (0, import_render_menu_action_children.renderMenuActionChildren)(defaultActions) : null,
      customActions ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { "data-dt-children-name": "LegendSeriesItems", children: (0, import_render_menu_action_children.renderMenuActionChildren)(customActions.props.children) }) : null
    ] })
  ] });
};
ChartLegendSeriesActions["displayName"] = "LegendSeriesActions";
