// @dynatrace/strato-components/core/dt-date/DtDateTime.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DtDateTime_exports = {};
__export(DtDateTime_exports, {
  DtDateTime: () => DtDateTime
});
module.exports = __toCommonJS(DtDateTime_exports);
var import_temporal_polyfill = require("temporal-polyfill");
var import_user_preferences = require("@dynatrace-sdk/user-preferences");
class DtDateTime {
  dateTime;
  constructor(dateTimeString, options) {
    const normalized = DtDateTime.normalizeRfc3339(dateTimeString);
    const defaultZone = options?.timezone || (0, import_user_preferences.getTimezone)() || "UTC";
    let parsedZdt = null;
    try {
      parsedZdt = import_temporal_polyfill.Temporal.Instant.from(normalized);
    } catch {
      parsedZdt = null;
    }
    if (parsedZdt) {
      this.dateTime = parsedZdt.toZonedDateTimeISO(defaultZone);
      return;
    }
    let pdt;
    try {
      pdt = import_temporal_polyfill.Temporal.PlainDateTime.from(normalized);
    } catch {
      throw new Error(`Invalid datetime string: ${dateTimeString}`);
    }
    this.dateTime = pdt.toZonedDateTime(defaultZone);
  }
  // --- Normalization --------------------------------------------------------
  /**
   * Normalize common RFC3339 variants:
   *  - lowercase 't' → 'T'
   *  - lowercase 'z' → 'Z'
   *  - space between date/time or underscore for Grail support → 'T'
   *
   * We only normalize the first separator between date and time (YYYY-MM-DD <sep> HH:mm…)
   * to avoid mangling valid strings that contain spaces elsewhere (e.g. annotations).
   * See: https://ijmacd.github.io/rfc3339-iso8601/ and https://www.rfc-editor.org/rfc/rfc3339.html#section-5.6
   */
  static normalizeRfc3339(input) {
    let s = input.trim();
    s = s.replace(/^\d{4}-\d{2}-\d{2}[ _]/, (m) => m.replace(/[ _]/, "T"));
    s = s.replace(
      /t/,
      (match, offset) => offset === s.indexOf("t") ? "T" : match
    );
    return s;
  }
  // --- Arithmetic (DST safe via ZonedDateTime) ------------------------------
  addDuration(duration) {
    const next = this.dateTime.add(duration);
    return DtDateTime.fromZdt(next);
  }
  addDays(days) {
    return this.addDuration({ days });
  }
  addWeeks(weeks) {
    return this.addDuration({ weeks });
  }
  // --- Output ---------------------------------------------------------------
  toString() {
    return this.dateTime.toString();
  }
  static fromZdt(zdt) {
    const obj = Object.create(DtDateTime.prototype);
    obj.dateTime = zdt;
    return obj;
  }
}
