// @dynatrace/strato-components/esm/charts/pie/utils/build-group-threshold.js
import { partition } from "lodash-es";
import { buildSlices } from "./build-slices.js";
import { THRESHOLD_TYPE } from "../types/default.js";
function buildGroupThreshold(data, grouping, colorPalette, unit) {
  const internalSlicesData = buildSlices(data, colorPalette, unit);
  const threshold = grouping?.threshold;
  if (threshold?.type === THRESHOLD_TYPE.NUMBER_OF_SLICES) {
    return calculateNumberOfSlicesGrouping(
      internalSlicesData,
      threshold,
      grouping?.name
    );
  }
  if (threshold?.type === THRESHOLD_TYPE.RELATIVE) {
    return calculateRelativeGrouping(
      internalSlicesData,
      threshold,
      grouping?.name
    );
  }
  if (threshold?.type === THRESHOLD_TYPE.ABSOLUTE) {
    return calculateAbsoluteGrouping(
      internalSlicesData,
      threshold,
      grouping?.name
    );
  }
  return internalSlicesData;
}
function calculateNumberOfSlicesGrouping(data, threshold, groupingName) {
  const slicesWithValue = data.filter((slice) => slice.value);
  const slicesToMerge = slicesWithValue.slice(threshold.value - 1);
  const remainingSlices = slicesWithValue.slice(0, threshold.value - 1);
  if (threshold.value === 0) {
    return [...slicesToMerge, ...remainingSlices];
  }
  return generateGroup(slicesToMerge, remainingSlices, groupingName);
}
function calculateRelativeGrouping(data, threshold, groupingName) {
  const slicesWithValue = data.filter((slice) => slice.value);
  const [slicesToMerge, remainingSlices] = partition(
    slicesWithValue,
    (slice) => slice.relativeValue < threshold.value
  );
  return generateGroup(slicesToMerge, remainingSlices, groupingName);
}
function calculateAbsoluteGrouping(data, threshold, groupingName) {
  const slicesWithValue = data.filter((slice) => slice.value);
  const [slicesToMerge, remainingSlices] = partition(
    slicesWithValue,
    (slice) => slice.value < threshold.value
  );
  return generateGroup(slicesToMerge, remainingSlices, groupingName);
}
function generateGroup(slicesToMerge, data, groupingName) {
  const isWithoutGrouping = !slicesToMerge.length || slicesToMerge.length === 1;
  if (isWithoutGrouping) {
    return [...data, ...slicesToMerge];
  }
  const groupIndexValue = slicesToMerge[0].id.split("-")[1];
  const groupId = groupingName ? groupingName + "-" + groupIndexValue : "other-" + groupIndexValue;
  const mergedSlice = slicesToMerge.reduce(
    (acc, curr) => ({
      ...acc,
      ...curr,
      category: groupingName ?? "Other",
      value: acc.value + curr.value,
      relativeValue: acc.relativeValue + curr.relativeValue,
      color: slicesToMerge[0].color,
      id: groupId
    }),
    {
      value: 0,
      relativeValue: 0
    }
  );
  return [...data, mergedSlice];
}
export {
  buildGroupThreshold
};
