// @dynatrace/strato-components/esm/charts/pie/utils/group-slices-thinner-than-pad-angle.js
import { getSlicePercentage } from "./plot-data.utils.js";
const groupSlicesThinnerThanPadAngle = (slices, padAngle) => {
  let slicesThinnerThanPadAngleValueSum = 0;
  let totalAbsoluteValueSum = 0;
  const newSlices = [];
  const padAngleDeg = padAngle * 180 / Math.PI;
  for (const slice of slices) {
    const sliceAngleDeg = slice.relativeValue / 100 * 360;
    if (sliceAngleDeg <= padAngleDeg) {
      slicesThinnerThanPadAngleValueSum += slice.value;
    } else {
      newSlices.push(slice);
    }
    totalAbsoluteValueSum += slice.value;
  }
  if (slicesThinnerThanPadAngleValueSum > 0) {
    const negativeSlicesAggregatedSliceColorToken = {
      Default: "transparent"
    };
    const negativeSlicesAggregatedSlice = {
      category: "",
      value: slicesThinnerThanPadAngleValueSum,
      color: negativeSlicesAggregatedSliceColorToken,
      id: "thinner-slices",
      relativeValue: getSlicePercentage(
        slicesThinnerThanPadAngleValueSum,
        totalAbsoluteValueSum
      )
    };
    newSlices.push(negativeSlicesAggregatedSlice);
  }
  return newSlices;
};
export {
  groupSlicesThinnerThanPadAngle
};
