// @dynatrace/strato-components/esm/charts/pie/utils/iterate-slots.js
import { Children } from "react";
import { Intent } from "../../../core/slots/Intent/Intent.js";
import { isSlot } from "../../../core/utils/is-slot.js";
import { ColorRule, isValidColorRule } from "../../core/slots/ColorRule.js";
import { ChartToolbar } from "../../core/slots/toolbar/ChartToolbar.js";
import { sanitizeProps } from "../../core/slots/toolbar/CompactToolbar.js";
import { DownloadData } from "../../core/slots/toolbar/DownloadData.js";
import { extractBehaviorTrackingProps } from "../../core/utils/extract-behavior-tracking-props.js";
import { Toolbar } from "../shared/slots/Toolbar.js";
import { DonutChartDownloadCSVSlot } from "../slots/DonutChartDownloadCSV.js";
import { Grouping as DonutChartGrouping } from "../slots/DonutChartGrouping.js";
import { DonutChartInner } from "../slots/DonutChartInner.js";
import { Labels as DonutChartLabels } from "../slots/DonutChartLabels.js";
import { Legend as DonutChartLegend } from "../slots/DonutChartLegend.js";
import { Tooltip as DonutChartTooltipSlot } from "../slots/DonutChartTooltip.js";
import { PieChartDownloadCSVSlot } from "../slots/PieChartDownloadCSV.js";
import { Grouping as PieChartGrouping } from "../slots/PieChartGrouping.js";
import { Labels as PieChartLabels } from "../slots/PieChartLabels.js";
import { Legend as PieChartLegend } from "../slots/PieChartLegend.js";
import { Tooltip as PieChartTooltipSlot } from "../slots/PieChartTooltip.js";
const iterateChartToolbarSlots = (children) => {
  let slots = {};
  Children.forEach(children, (childInteractions) => {
    if (isSlot(childInteractions, DownloadData)) {
      slots = { ...slots, downloadData: { enabled: true } };
    }
  });
  return slots;
};
function iterateSlots(children) {
  let slots = {
    labels: {},
    grouping: {},
    legend: {},
    toolbar: void 0,
    innerElement: void 0,
    colorRules: [],
    tooltip: void 0
  };
  let intents;
  Children.map(children, (child) => {
    if (isSlot(child, DonutChartLabels) || isSlot(child, PieChartLabels)) {
      slots = { ...slots, labels: child.props };
    } else if (isSlot(child, PieChartGrouping) || isSlot(child, DonutChartGrouping)) {
      slots = { ...slots, grouping: child.props };
    } else if (isSlot(child, PieChartLegend) || isSlot(child, DonutChartLegend)) {
      slots = { ...slots, legend: child.props };
    } else if (isSlot(child, Toolbar)) {
      slots = {
        ...slots,
        toolbar: {
          ...slots.toolbar,
          ...sanitizeProps(child.props),
          ...extractBehaviorTrackingProps(child.props)
        }
      };
    } else if (isSlot(child, ChartToolbar)) {
      const { children: children2, ...toolbarProps } = child.props;
      const toolbarSlots = iterateChartToolbarSlots(children2);
      const toolbar = { ...toolbarProps, ...toolbarSlots };
      slots = { ...slots, toolbar };
    } else if (isSlot(child, DonutChartDownloadCSVSlot) || isSlot(child, PieChartDownloadCSVSlot)) {
      slots = {
        ...slots,
        toolbar: {
          ...slots.toolbar,
          downloadData: { enabled: true }
        }
      };
    } else if (isSlot(child, Intent)) {
      intents = intents ? [...intents, child.props] : [child.props];
    } else if (isSlot(child, DonutChartInner)) {
      const inner = child.props.children;
      slots = { ...slots, innerElement: inner };
    } else if (isSlot(child, ColorRule)) {
      isValidColorRule(child.props) && slots.colorRules.push(child.props);
    } else if (isSlot(child, PieChartTooltipSlot) || isSlot(child, DonutChartTooltipSlot)) {
      slots = {
        ...slots,
        tooltip: {
          hidden: child.props.hidden ?? false,
          template: child.props.children
        }
      };
    }
  });
  if (intents?.length) {
    slots = {
      ...slots,
      toolbar: {
        ...slots.toolbar,
        intents
      }
    };
  }
  return slots;
}
export {
  iterateSlots
};
