// @dynatrace/strato-components/esm/charts/pie/utils/pie-config.js
import { isUndefined, merge } from "lodash-es";
import { DEFAULT_TRUNCATION_MODE } from "../../core/constants.js";
import { isValidColorRule } from "../../core/slots/ColorRule.js";
import {
  isCustomColorPalette
} from "../../core/types/color-palette.js";
import { defaultUnitFormatter } from "../../core/utils/defaultUnitFormatter.js";
import { isValidColorPalette } from "../../core/utils/is-valid-color-palette.js";
import { isValidLegendPosition } from "../../core/utils/is-valid-legend-position.js";
import { THRESHOLD_TYPE_OPTION } from "../constants.js";
import { defaultConfig } from "../types/default.js";
import { isValidThresholdType } from "../utils/is-valid-threshold-type.js";
function getColorPaletteConfiguration(contextColorPalette, propsColorPalette) {
  const colorPalette = propsColorPalette ?? contextColorPalette ?? defaultConfig.colorPalette;
  return isCustomColorPalette(colorPalette) || isValidColorPalette(colorPalette) ? colorPalette : defaultConfig.colorPalette;
}
function getGroupingConfiguration(contextGrouping, slotsGrouping) {
  let grouping = merge(
    {},
    defaultConfig.grouping,
    contextGrouping,
    slotsGrouping
  );
  grouping = {
    ...grouping,
    threshold: {
      ...grouping.threshold,
      type: isValidThresholdType(grouping.threshold.type) ? grouping.threshold.type : THRESHOLD_TYPE_OPTION.Relative
    }
  };
  return grouping;
}
function getLegendConfiguration(contextLegendConfig, slotsLegendConfig) {
  let legend = merge(
    {},
    defaultConfig.legend,
    contextLegendConfig,
    slotsLegendConfig
  );
  const position = legend.position && isValidLegendPosition(legend.position) ? legend.position : defaultConfig.legend.position;
  legend = {
    ...legend,
    position
  };
  return legend;
}
const validateToolbarContextConfig = (contextConfig) => !isUndefined(contextConfig) && contextConfig.downloadData?.enabled;
const getToolbarConfiguration = (contextConfig, slots) => {
  const hasValidToolbarConfig = validateToolbarContextConfig(contextConfig) || !isUndefined(slots);
  return hasValidToolbarConfig ? merge({}, contextConfig, slots) : void 0;
};
function buildPieChartConfig(props, contextConfig, slots) {
  const colorPalette = getColorPaletteConfiguration(
    contextConfig.colorPalette,
    props.colorPalette
  );
  const grouping = getGroupingConfiguration(
    contextConfig.grouping,
    slots.grouping
  );
  const legend = getLegendConfiguration(contextConfig.legend, slots.legend);
  const formatter = props.formatter ?? contextConfig.formatter ?? defaultUnitFormatter(props.data.unit, 0);
  const truncationMode = props.truncationMode ?? contextConfig.truncationMode ?? DEFAULT_TRUNCATION_MODE;
  const toolbar = getToolbarConfiguration(contextConfig.toolbar, slots.toolbar);
  const seriesActions = props.seriesActions ?? contextConfig.seriesActions ?? void 0;
  const colorRules = (contextConfig.colorRules || []).filter(isValidColorRule).concat(slots.colorRules || []);
  return {
    height: props.height ?? defaultConfig.height,
    colorPalette,
    labels: merge({}, defaultConfig.labels, contextConfig.labels, slots.labels),
    grouping,
    legend,
    formatter,
    truncationMode,
    toolbar,
    innerElement: slots.innerElement,
    seriesActions,
    colorRules,
    tooltip: slots.tooltip ?? contextConfig.tooltip
  };
}
function deserializeConfig(serializedConfig) {
  try {
    return merge({}, defaultConfig, JSON.parse(serializedConfig));
  } catch {
    return defaultConfig;
  }
}
export {
  buildPieChartConfig,
  deserializeConfig,
  getColorPaletteConfiguration,
  getGroupingConfiguration,
  getLegendConfiguration
};
