// @dynatrace/strato-components/esm/charts/meter-bar/multi-meter-bar/MultiMeterBarChart.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import { forwardRef, useRef } from "react";
import { MultiMeterBarChartLegend } from "./components/MultiMeterBarChartLegend.js";
import { MultiMeterBarChartPlotArea } from "./components/MultiMeterBarChartPlotArea.js";
import { MULTIMETER_BAR_CHART_ARIA_LABEL } from "./constants.js";
import { useMultiMeterBarChartConfig } from "./hooks/useMultiMeterBarChartConfig.js";
import { MultiMeterBarChartProviders } from "./providers/MultiMeterBarChartProviders.js";
import { Label } from "./slots/Label.js";
import { Legend } from "./slots/Legend.js";
import { Max } from "./slots/Max.js";
import { Min } from "./slots/Min.js";
import { Segment } from "./slots/Segment.js";
import { Tooltip } from "./slots/Tooltip.js";
import { Value } from "./slots/Value.js";
import { useBehavioralTrackingProps } from "../../../core/hooks/useBehavioralTrackingProps.js";
import { Intent } from "../../../core/slots/Intent/Intent.js";
import { defaultChartSize } from "../../../core/styles/defaultChartValues.sty.js";
import { ColorRule } from "../../core/slots/ColorRule.js";
import { ChartType } from "../../core/types/chart-types.js";
import { MeterBarChartFormattedValue } from "../shared/components/formatted-value/MeterBarChartFormattedValue.js";
import { MeterBarLayout } from "../shared/components/meter-bar-layout/MeterBarLayout.js";
import { MIN_DEFAULT_VALUE, SIZE_DEFAULT } from "../shared/constants.js";
import { Toolbar } from "../shared/slots/Toolbar.js";
const COMPONENT_NAME = "MultiMeterBarChart";
const MultiMeterBarChartRoot = forwardRef((props, forwardedRef) => {
  const {
    width,
    height,
    className,
    style,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "data-testid": dataTestId = "multi-meter-bar-chart-root",
    seriesActions,
    prefixIcon
  } = props;
  const [behavioralTrackingProps] = useBehavioralTrackingProps(props);
  const config = useMultiMeterBarChartConfig(props);
  const {
    colorPalette,
    size = SIZE_DEFAULT,
    min = MIN_DEFAULT_VALUE,
    max,
    loading,
    formatter,
    unit,
    labelSlot,
    valueSlot,
    minSlot,
    maxSlot,
    segments,
    legend,
    tooltip,
    totalValue,
    colorRules,
    toolbar
  } = config;
  const multiMeterBarChartRef = useRef(null);
  const consumerConfig = {
    type: ChartType.MultiMeterBar,
    colorPalette,
    size,
    min,
    max,
    formatter,
    unit,
    tooltip: { enabled: !tooltip.hidden, variant: tooltip.variant },
    legend,
    loading
  };
  const label = labelSlot ?? "";
  return /* @__PURE__ */ jsx(
    MultiMeterBarChartProviders,
    {
      colorPalette,
      formatter,
      unit,
      segments,
      colorRules,
      toolbar,
      loading,
      seriesActions,
      children: /* @__PURE__ */ jsxs(
        MeterBarLayout,
        {
          ...behavioralTrackingProps,
          width,
          height,
          ref: multiMeterBarChartRef,
          size,
          className: clsx(defaultChartSize, className),
          style,
          "data-dtrum-mask": dataDtrumMask,
          "data-dtrum-allow": dataDtrumAllow,
          "data-testid": dataTestId,
          "aria-label": MULTIMETER_BAR_CHART_ARIA_LABEL,
          "data-dt-component": COMPONENT_NAME,
          consumerConfig,
          userRef: forwardedRef,
          children: [
            prefixIcon && /* @__PURE__ */ jsx(MeterBarLayout.Icon, { children: prefixIcon }),
            labelSlot && /* @__PURE__ */ jsx(MeterBarLayout.Label, { children: labelSlot }),
            valueSlot && /* @__PURE__ */ jsx(MeterBarLayout.Value, { children: /* @__PURE__ */ jsx(
              MeterBarChartFormattedValue,
              {
                value: totalValue,
                total: max - min,
                auto: valueSlot?.auto,
                children: valueSlot?.value
              }
            ) }),
            /* @__PURE__ */ jsx(MeterBarLayout.Chart, { children: /* @__PURE__ */ jsx(
              MultiMeterBarChartPlotArea,
              {
                label,
                max,
                min,
                tooltipConfig: tooltip
              }
            ) }),
            minSlot && /* @__PURE__ */ jsx(MeterBarLayout.Min, { children: /* @__PURE__ */ jsx(MeterBarChartFormattedValue, { value: min, auto: minSlot?.auto, children: minSlot?.value }) }),
            maxSlot && /* @__PURE__ */ jsx(MeterBarLayout.Max, { children: /* @__PURE__ */ jsx(MeterBarChartFormattedValue, { value: max, auto: maxSlot?.auto, children: maxSlot?.value }) }),
            legend && /* @__PURE__ */ jsx(MeterBarLayout.Legend, { children: /* @__PURE__ */ jsx(MultiMeterBarChartLegend, {}) })
          ]
        }
      )
    }
  );
});
MultiMeterBarChartRoot.displayName = "_MultiMeterBarChart";
const MultiMeterBarChart = Object.assign(MultiMeterBarChartRoot, {
  Label,
  Legend,
  Max,
  Min,
  Segment,
  Value,
  Tooltip,
  ColorRule,
  Intent,
  Toolbar
});
export {
  MultiMeterBarChart
};
