// @dynatrace/strato-components/esm/forms/forms-core/hooks/useFormFieldValidation.js
import { jsx } from "react/jsx-runtime";
import {
  useCallback,
  useMemo,
  useEffect,
  useRef
} from "react";
import { FormattedMessage } from "react-intl";
import { useFormFieldMessagesContext } from "./useFormFieldMessagesContext.js";
import { useId } from "../../../core/hooks/useId.js";
import { Text } from "../../../typography/text/Text.js";
import { FormFieldMessagesAction } from "../state/formfield-messages-state-reducer.js";
function getErrorType(validityState) {
  if (!validityState) {
    return null;
  }
  if (validityState.badInput) {
    return "badInput";
  }
  if (validityState.patternMismatch) {
    return "patternMismatch";
  }
  if (validityState.rangeOverflow) {
    return "rangeOverflow";
  }
  if (validityState.rangeUnderflow) {
    return "rangeUnderflow";
  }
  if (validityState.stepMismatch) {
    return "stepMismatch";
  }
  if (validityState.tooLong) {
    return "tooLong";
  }
  if (validityState.tooShort) {
    return "tooShort";
  }
  if (validityState.typeMismatch) {
    return "typeMismatch";
  }
  if (validityState.valueMissing) {
    return "valueMissing";
  }
  return null;
}
function getI18nErrorMessage(key, ref) {
  switch (key) {
    case "badInput":
      return /* @__PURE__ */ jsx(
        FormattedMessage,
        {
          id: "/cwiOkubC69MZmRu",
          defaultMessage: "Please provide input in a valid format.",
          description: "Error message for input that the browser is unable to convert, e.g. string for a number input."
        }
      );
    case "patternMismatch":
      return /* @__PURE__ */ jsx(
        FormattedMessage,
        {
          id: "qPeX7AgKbKzBFqGZ",
          defaultMessage: "Please match the format requested.",
          description: "Error message for input that doesn't follow the requested pattern."
        }
      );
    case "rangeOverflow":
      return /* @__PURE__ */ jsx(
        FormattedMessage,
        {
          id: "14mVeqEFpqD2fpev",
          defaultMessage: "Value must be less or equal to {max}.",
          description: "Error message for input that value is higher then the max.",
          values: { max: ref?.getAttribute("max") }
        }
      );
    case "rangeUnderflow":
      return /* @__PURE__ */ jsx(
        FormattedMessage,
        {
          id: "5XoVjJUQ5b86MLW+",
          defaultMessage: "Value must be higher or equal to {min}.",
          description: "Error message for input that value is lower then the max.",
          values: { min: ref?.getAttribute("min") }
        }
      );
    case "stepMismatch":
      return /* @__PURE__ */ jsx(
        FormattedMessage,
        {
          id: "MYfVMxPsAgOn8JSP",
          defaultMessage: "Value must be a multiple of {step}, starting from {min}.",
          description: "Only values which are a whole number of steps from the step base are valid. The step base is min if specified, value otherwise, or 0 if neither is provided.",
          values: {
            step: ref?.getAttribute("step"),
            min: ref?.getAttribute("min") ?? "0"
          }
        }
      );
    case "tooLong":
      return /* @__PURE__ */ jsx(
        FormattedMessage,
        {
          id: "JOIA3jQfIIkgnW04",
          defaultMessage: "Please shorten this text to {maxLength} characters or less.",
          description: "Error message for input that is shorter than the maxLength.",
          values: { maxLength: ref?.getAttribute("maxLength") }
        }
      );
    case "tooShort":
      return /* @__PURE__ */ jsx(
        FormattedMessage,
        {
          id: "sJeuQTGrzcMvd6fY",
          defaultMessage: "Please lengthen this text to {minLength} characters or more.",
          description: "Error message for input that is shorter than the minLength.",
          values: { minLength: ref?.getAttribute("minLength") }
        }
      );
    case "typeMismatch":
      return /* @__PURE__ */ jsx(
        FormattedMessage,
        {
          id: "fLfhKO8cXlo3vW5M",
          defaultMessage: "Please provide a value that matches the requested type ({type}).",
          description: "Error message for input that doesn't match the expected type.",
          values: { type: ref?.getAttribute("type") }
        }
      );
    case "valueMissing":
      return /* @__PURE__ */ jsx(
        FormattedMessage,
        {
          id: "laIylhVFbIvEZrxt",
          defaultMessage: "Please fill in this field.",
          description: "Error message for missing input."
        }
      );
    case "customError":
    case "valid":
    default:
      return null;
  }
}
function useFormFieldValidationAriaProps(ariaLabelingProps = {}) {
  const { state } = useFormFieldMessagesContext();
  const _ariaProps = useMemo(() => {
    let messageIdsString;
    let errorMessageIdsString;
    if (state.customMessages.length > 0 || state.formControlMessages.length > 0) {
      const messages = [...state.customMessages, ...state.formControlMessages];
      errorMessageIdsString = messages.filter((msg) => msg.variant === "error").map((msg) => msg.id).join(" ");
      messageIdsString = messages.map((msg) => msg.id).join(" ");
    }
    return {
      "aria-errormessage": state.showError ? errorMessageIdsString : void 0,
      "aria-invalid": state.showError,
      "aria-describedby": messageIdsString
    };
  }, [state.customMessages, state.formControlMessages, state.showError]);
  const ariaProps = getCombinedAriaProps(_ariaProps, ariaLabelingProps);
  return ariaProps;
}
function useFormFieldValidation(ref, validationDeps = {}, ariaLabelingProps = {}, controlId) {
  const { state, dispatch } = useFormFieldMessagesContext();
  const errorId = useId();
  const ariaProps = useFormFieldValidationAriaProps(ariaLabelingProps);
  const formControlMessagesRef = useRef(state.formControlMessages);
  formControlMessagesRef.current = state.formControlMessages;
  const updateErrorMessages = useCallback(() => {
    const errorType = getErrorType(ref.current?.validity);
    formControlMessagesRef.current.forEach((message) => {
      if (message.validityStateKey !== errorType && message.validityStateKey !== "customError") {
        dispatch?.({
          type: FormFieldMessagesAction.REMOVEMESSAGE,
          payload: {
            id: message.id,
            messageType: "formcontrol",
            controlId
          }
        });
      }
    });
    if (errorType) {
      dispatch?.({
        type: FormFieldMessagesAction.ADDMESSAGE,
        payload: {
          id: errorId,
          message: /* @__PURE__ */ jsx(Text, { textStyle: "small", as: "p", children: getI18nErrorMessage(errorType, ref.current) }),
          validityStateKey: errorType,
          messageType: "formcontrol",
          variant: "error",
          controlId
        }
      });
    }
  }, [controlId, dispatch, errorId, ref]);
  const onInvalid = useCallback(
    (event) => {
      updateErrorMessages();
      dispatch?.({
        type: FormFieldMessagesAction.SETISTOUCHED,
        payload: { isTouched: true }
      });
      event.preventDefault();
    },
    [dispatch, updateErrorMessages]
  );
  const onBlur = useCallback(() => {
    ref.current?.reportValidity();
    dispatch?.({
      type: FormFieldMessagesAction.SETISTOUCHED,
      payload: { isTouched: true }
    });
  }, [dispatch, ref]);
  const validationDepsKey = JSON.stringify(validationDeps);
  useEffect(() => {
    updateErrorMessages();
  }, [validationDepsKey, updateErrorMessages]);
  return {
    ariaProps,
    onInvalid,
    onChange: updateErrorMessages,
    onBlur
  };
}
function getCombinedAriaProps(validationAriaProps, labelingProps) {
  const {
    "aria-describedby": validationDescription,
    ...remainingValidationProps
  } = validationAriaProps;
  const { "aria-describedby": propDescription, ...remainingLabelingProps } = labelingProps;
  if (validationDescription && propDescription) {
    return {
      ...remainingLabelingProps,
      ...remainingValidationProps,
      "aria-describedby": (validationDescription + " " + propDescription).trim()
    };
  }
  return {
    ...remainingLabelingProps,
    ...remainingValidationProps,
    "aria-describedby": validationDescription ?? propDescription
  };
}
export {
  getCombinedAriaProps,
  getErrorType,
  getI18nErrorMessage,
  useFormFieldValidation,
  useFormFieldValidationAriaProps
};
