// @dynatrace/strato-components/charts/xy-chart/store/reducer-explorer.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var reducer_explorer_exports = {};
__export(reducer_explorer_exports, {
  reduceExploreCancelAll: () => reduceExploreCancelAll,
  reduceExploreCancelHighlight: () => reduceExploreCancelHighlight,
  reduceExploreCancelSelect: () => reduceExploreCancelSelect,
  reduceExplorePin: () => reduceExplorePin,
  reduceExploreReset: () => reduceExploreReset,
  reduceExploreSelect: () => reduceExploreSelect,
  reduceExploreUnPin: () => reduceExploreUnPin
});
module.exports = __toCommonJS(reducer_explorer_exports);
function reduceExploreReset(prev) {
  return {
    ...prev,
    highlightedSeriesId: void 0,
    select: void 0
  };
}
function reduceExplorePin(highlightedSeriesId) {
  return (prev) => ({
    ...prev,
    highlightedSeriesId
  });
}
function reduceExploreUnPin(prev) {
  return {
    ...prev,
    select: void 0,
    highlightedSeriesId: void 0
  };
}
function reduceExploreSelect(select) {
  return (prev) => ({
    ...prev,
    select: {
      ...prev.select,
      ...select
    },
    highlightedSeriesId: void 0
  });
}
function reduceExploreCancelSelect(prev) {
  return {
    ...prev,
    select: void 0
  };
}
function reduceExploreCancelHighlight(prev) {
  return {
    ...prev,
    highlightedSeriesId: void 0
  };
}
function reduceExploreCancelAll(prev) {
  return {
    ...prev,
    select: void 0,
    highlightedSeriesId: void 0
  };
}
