// @dynatrace/strato-components/charts/xy-chart/store/selectors.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var selectors_exports = {};
__export(selectors_exports, {
  useAnnotationsState: () => useAnnotationsState,
  useCurrentDomainState: () => useCurrentDomainState,
  useFilteredSeriesIdsState: () => useFilteredSeriesIdsState,
  useHoveredThresholdState: () => useHoveredThresholdState,
  useInteractionModeState: () => useInteractionModeState,
  useSelectState: () => useSelectState,
  useSelector: () => useSelector,
  useSetState: () => useSetState,
  useXYChartHighlightedSeriesIdsState: () => useXYChartHighlightedSeriesIdsState,
  useZoomState: () => useZoomState
});
module.exports = __toCommonJS(selectors_exports);
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_selector = require("../../../core/store/selector.js");
var import_XYChartStore_context = require("../context/XYChartStore.context.js");
const useSelector = (selector, compareFn) => (0, import_selector.useStoreSelector)(import_XYChartStore_context.XYChartStoreContext, selector, compareFn);
const useSetState = () => {
  const { setState } = (0, import_react.useContext)(import_XYChartStore_context.XYChartStoreContext);
  return setState;
};
const useXYChartHighlightedSeriesIdsState = () => useSelector((state) => state.highlightedSeriesId, import_lodash_es.isEqual);
const useInteractionModeState = () => useSelector((state) => state.interactionMode);
const useZoomState = () => useSelector((state) => state.zoom, import_lodash_es.isEqual);
const useCurrentDomainState = () => useSelector((state) => state.currentDomain, import_lodash_es.isEqual);
const useSelectState = () => useSelector((state) => state.select);
const useHoveredThresholdState = () => useSelector((state) => state.thresholds.hoveredThresholdId);
const useAnnotationsState = () => useSelector((state) => state.annotations);
const areSetsEqual = (a, b) => {
  if (a === b) {
    return true;
  }
  if (!a || !b) {
    return false;
  }
  if (a.size !== b.size) {
    return false;
  }
  for (const value of a) {
    if (!b.has(value)) {
      return false;
    }
  }
  return true;
};
const useFilteredSeriesIdsState = () => useSelector((state) => state.filteredSeriesIds, areSetsEqual);
