// @dynatrace/strato-components/esm/navigation/menu/Content.js
import { jsx } from "react/jsx-runtime";
import {
  Content as RadixContent,
  Portal as RadixPortal
} from "@radix-ui/react-dropdown-menu";
import {
  forwardRef,
  useCallback
} from "react";
import { useMenuContext } from "./hooks/useMenuContext.js";
import * as styles from "./Menu.sty.js";
import { useOverlayStacking } from "../../core/hooks/useOverlayStacking.js";
const SIDE_OFFSET = 4;
const Content = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const {
    side = "bottom",
    alignment = "start",
    onCloseAutoFocus,
    avoidCollisions = true,
    _sideOffset = SIDE_OFFSET,
    _onTabKeyPressed,
    children,
    collisionBoundary,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    ...remainingProps
  } = props;
  const tabKeyHandler = useCallback(
    (e) => {
      if (_onTabKeyPressed && e.key === "Tab") {
        _onTabKeyPressed();
      }
    },
    [_onTabKeyPressed]
  );
  const { style } = useOverlayStacking(void 0, "overlay");
  const { setIsOpen } = useMenuContext();
  const handleOnKeyDown = useCallback(
    (e) => {
      if (e.key === "Escape") {
        setIsOpen(false);
        e.stopPropagation();
      }
    },
    [setIsOpen]
  );
  return /* @__PURE__ */ jsx(RadixPortal, { children: /* @__PURE__ */ jsx(
    RadixContent,
    {
      ...remainingProps,
      "data-testid": dataTestId,
      ref: forwardedRef,
      side,
      align: alignment,
      onKeyUp: tabKeyHandler,
      onCloseAutoFocus,
      sideOffset: _sideOffset,
      avoidCollisions,
      className: styles.content,
      style: { zIndex: style.zIndex },
      onKeyDown: handleOnKeyDown,
      collisionBoundary,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      children
    }
  ) });
});
Content.displayName = "Menu.Content";
export {
  Content
};
