// @dynatrace/strato-components/esm/navigation/menu/Intent.js
import { jsx, jsxs } from "react/jsx-runtime";
import {
  forwardRef,
  useMemo
} from "react";
import {
  getIntentLink
} from "@dynatrace-sdk/navigation";
import { getIntentSlots } from "./item-slots.js";
import { Item } from "./Item.js";
import { Prefix } from "./Prefix.js";
import {
  useIntent,
  useIntentWithResponse
} from "../../core/hooks/intent.js";
import { AppIcon } from "../../layouts/app-header/app-icon/AppIcon.js";
function isWithResponse(props) {
  if ("onResponse" in props) {
    return typeof props.onResponse === "function";
  }
  return false;
}
function getIntentIcon(icon, intentIcon) {
  if (!icon && !intentIcon) {
    return null;
  }
  if (icon) {
    return icon;
  }
  if (intentIcon) {
    return /* @__PURE__ */ jsx(Prefix, { children: /* @__PURE__ */ jsx(AppIcon, { src: intentIcon, alt: "" }) });
  }
}
const getHandleSelect = (intent, onSelect, onResponse) => async (e) => {
  onSelect?.(e);
  try {
    const response = await intent.send();
    onResponse(response);
  } catch {
    onResponse({
      status: "error",
      error: new Error("Error sending intent with response")
    });
  }
};
const IntentWithoutResponse = /* @__PURE__ */ forwardRef((props, ref) => {
  const {
    onResponse,
    onSelect,
    payload,
    options,
    children,
    disabled,
    ...rest
  } = props;
  const intent = useIntent({
    payload,
    options
  });
  const slots = getIntentSlots(children);
  const intentIcon = getIntentIcon(slots.prefix, intent.iconURL);
  const link = useMemo(() => {
    return getIntentLink(
      payload,
      options.recommendedAppId ?? "",
      options.recommendedIntentId ?? ""
    );
  }, [options.recommendedAppId, options.recommendedIntentId, payload]);
  return /* @__PURE__ */ jsxs(
    Item,
    {
      ...rest,
      as: "a",
      href: disabled ? void 0 : link,
      ref,
      disabled,
      onSelect: (e) => {
        e.preventDefault();
        intent.send();
        onSelect?.(e);
      },
      children: [
        intentIcon,
        slots.freeform
      ]
    }
  );
});
const IntentWithResponse = /* @__PURE__ */ forwardRef((props, ref) => {
  const {
    onSelect,
    onResponse,
    payload,
    options,
    children,
    disabled,
    ...rest
  } = props;
  const intent = useIntentWithResponse({
    payload,
    options
  });
  const slots = getIntentSlots(children);
  const intentIcon = getIntentIcon(slots.prefix, intent.iconURL);
  const handleSelect = getHandleSelect(intent, onSelect, onResponse);
  return /* @__PURE__ */ jsxs(Item, { ...rest, ref, disabled, onSelect: handleSelect, children: [
    intentIcon,
    slots.freeform
  ] });
});
const Intent = /* @__PURE__ */ forwardRef(
  (props, ref) => {
    if (isWithResponse(props)) {
      return /* @__PURE__ */ jsx(IntentWithResponse, { ...props, ref });
    }
    return /* @__PURE__ */ jsx(IntentWithoutResponse, { ...props, ref });
  }
);
Intent.displayName = "Menu.Intent";
export {
  Intent,
  getHandleSelect,
  getIntentIcon
};
