// @dynatrace/strato-components/esm/navigation/menu/item-slots.js
import { Children } from "react";
import { isElement } from "react-is";
import { ItemTooltip } from "./ItemTooltip.js";
import { Prefix } from "./Prefix.js";
import { Suffix } from "./Suffix.js";
function getItemSlots(children) {
  const slots = {
    tooltip: null,
    prefix: null,
    suffix: null,
    freeform: []
  };
  Children.forEach(children, (child) => {
    if (isElement(child)) {
      switch (child.type) {
        case ItemTooltip:
          slots.tooltip = child;
          break;
        case Prefix:
          slots.prefix = child;
          break;
        case Suffix:
          slots.suffix = child;
          break;
        default:
          slots.freeform.push(child);
      }
    } else {
      slots.freeform.push(child);
    }
  });
  return slots;
}
function getSubTriggerSlots(children) {
  const slots = {
    prefix: null,
    freeform: []
  };
  Children.forEach(children, (child) => {
    if (isElement(child)) {
      switch (child.type) {
        case ItemTooltip:
          break;
        case Prefix:
          slots.prefix = child;
          break;
        case Suffix:
          break;
        default:
          slots.freeform.push(child);
      }
    } else {
      slots.freeform.push(child);
    }
  });
  return slots;
}
const getIntentSlots = getSubTriggerSlots;
export {
  getIntentSlots,
  getItemSlots,
  getSubTriggerSlots
};
