// @dynatrace/strato-components/esm/navigation/menu/Item.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { Item as RadixItem } from "@radix-ui/react-dropdown-menu";
import clsx from "clsx";
import {
  forwardRef,
  useRef,
  useState
} from "react";
import { useIsMenuItemHovered } from "./hooks/useIsMenuItemHovered.js";
import { getItemSlots } from "./item-slots.js";
import { MenuItemTooltipWrapper } from "./ItemTooltip.js";
import * as styles from "./Menu.sty.js";
import { useFocusRing } from "../../core/styles/useFocusRing.js";
import { mergeProps as _mergeProps } from "../../core/utils/merge-props.js";
import { Flex } from "../../layouts/flex/Flex.js";
import { useTooltipState } from "../../overlays/hooks/useTooltipState.js";
import { textStyle as textStyleCSS } from "../../styles/textStyle.sty.js";
const Item = /* @__PURE__ */ forwardRef(
  ({
    children,
    disabled = false,
    onSelect,
    as = "div",
    "aria-label": ariaLabel,
    "aria-labelledby": ariaLabelledBy,
    "aria-describedby": ariaDescribedBy,
    "aria-details": ariaDetails,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    ...remainingProps
  }, forwardedRef) => {
    const slots = getItemSlots(children);
    const {
      isOpen,
      open: openTooltip,
      close: closeTooltip
    } = useTooltipState({
      defaultOpen: false
    });
    const [tooltipTriggerRef, setTooltipTriggerRef] = useState(null);
    const { focusProps, focusClassName } = useFocusRing({
      disabled
    });
    const isTooltipHovered = useRef(false);
    const { isItemHovered, hoverProps } = useIsMenuItemHovered();
    return /* @__PURE__ */ jsxs(Fragment, { children: [
      /* @__PURE__ */ jsx(
        RadixItem,
        {
          "data-testid": dataTestId,
          ref: forwardedRef,
          ..._mergeProps(focusProps, hoverProps, {
            onClick: (event) => {
              if (disabled) {
                event.stopPropagation();
                event.preventDefault();
                return;
              }
              const nativeEvent = event.nativeEvent;
              if (nativeEvent.button !== 0 || nativeEvent.metaKey || nativeEvent.ctrlKey || nativeEvent.shiftKey) {
                return;
              }
              onSelect(event);
            },
            onFocus: () => slots.tooltip && openTooltip(true),
            onBlur: () => slots.tooltip && // Ensures isTooltipHovered state is set correctly when the tooltip is hovered.
            // Otherwise, the tooltip would close before the onMouseLeave event of the tooltip triggers
            // and the isTooltipHovered state would never be set.
            requestAnimationFrame(() => {
              if (!isTooltipHovered.current) {
                closeTooltip();
              }
            }),
            className: consumerClassName,
            style: consumerStyle
          }),
          "aria-label": ariaLabel,
          "aria-labelledby": ariaLabelledBy,
          "aria-describedby": ariaDescribedBy,
          "aria-details": ariaDetails,
          asChild: true,
          children: /* @__PURE__ */ jsxs(
            Flex,
            {
              ...remainingProps,
              className: clsx(
                textStyleCSS({ textStyle: "base-emphasized" }),
                !isItemHovered && focusClassName,
                styles.item({ disabled })
              ),
              disabled,
              role: "menuitem",
              "aria-disabled": disabled,
              ref: setTooltipTriggerRef,
              as,
              children: [
                slots.prefix,
                /* @__PURE__ */ jsx("div", { className: styles.itemChildren, children: slots.freeform }),
                slots.suffix
              ]
            }
          )
        }
      ),
      isOpen && /* @__PURE__ */ jsx(
        MenuItemTooltipWrapper,
        {
          triggerElement: tooltipTriggerRef,
          onMouseEnter: () => {
            isTooltipHovered.current = true;
          },
          onMouseLeave: () => {
            isTooltipHovered.current = false;
            closeTooltip();
          },
          children: slots.tooltip
        }
      )
    ] });
  }
);
Item.displayName = "Menu.Item";
export {
  Item
};
