// @dynatrace/strato-components/esm/navigation/menu/ItemTooltip.js
import { jsx } from "react/jsx-runtime";
import { forwardRef, useState } from "react";
import { usePopper } from "react-popper";
import * as menuStyles from "./Menu.sty.js";
import { ALIGN_OFFSET } from "./SubContent.js";
const MenuItemTooltipWrapper = ({
  triggerElement,
  children,
  onMouseEnter,
  onMouseLeave
}) => {
  const [menuTooltipRefRef, setMenuTooltipRefRef] = useState(null);
  const { styles, attributes } = usePopper(triggerElement, menuTooltipRefRef, {
    placement: "right-start",
    modifiers: [
      {
        name: "offset",
        options: {
          offset: [ALIGN_OFFSET, 0]
        }
      },
      {
        name: "flip",
        options: {
          fallbackPlacements: ["left-start"]
        }
      }
    ]
  });
  return /* @__PURE__ */ jsx(
    "div",
    {
      ref: setMenuTooltipRefRef,
      style: styles.popper,
      className: menuStyles.tooltipWrapper,
      onMouseEnter,
      onMouseLeave,
      ...attributes.popper,
      children
    }
  );
};
MenuItemTooltipWrapper.displayName = "MenuItemTooltipWrapper";
const ItemTooltip = forwardRef(
  ({ children, "data-testid": dataTestId }, forwardedRef) => children ? /* @__PURE__ */ jsx(
    "div",
    {
      className: menuStyles.tooltip,
      ref: forwardedRef,
      "data-testid": dataTestId,
      children
    }
  ) : null
);
ItemTooltip.displayName = "Menu.ItemTooltip";
export {
  ItemTooltip,
  MenuItemTooltipWrapper
};
