// @dynatrace/strato-components/esm/navigation/menu/Link.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { Item as RadixItem } from "@radix-ui/react-dropdown-menu";
import clsx from "clsx";
import {
  forwardRef,
  useState
} from "react";
import { useIsMenuItemHovered } from "./hooks/useIsMenuItemHovered.js";
import { getItemSlots } from "./item-slots.js";
import { MenuItemTooltipWrapper } from "./ItemTooltip.js";
import * as styles from "./Menu.sty.js";
import { useFocusRing } from "../../core/styles/useFocusRing.js";
import { mergeProps as _mergeProps } from "../../core/utils/merge-props.js";
import { Flex } from "../../layouts/flex/Flex.js";
import { useTooltipState } from "../../overlays/hooks/useTooltipState.js";
import { textStyle as textStyleCSS } from "../../styles/textStyle.sty.js";
const Link = /* @__PURE__ */ forwardRef(
  ({
    children,
    disabled = false,
    as = "a",
    "aria-label": ariaLabel,
    "aria-labelledby": ariaLabelledBy,
    "aria-describedby": ariaDescribedBy,
    "aria-details": ariaDetails,
    className: consumerClassName,
    style: consumerStyle,
    ...remainingProps
  }, forwardedRef) => {
    const slots = getItemSlots(children);
    const {
      isOpen,
      open: openTooltip,
      close: closeTooltip
    } = useTooltipState({
      defaultOpen: false
    });
    const [tooltipTriggerRef, setTooltipTriggerRef] = useState(null);
    const { focusProps, focusClassName } = useFocusRing({
      disabled
    });
    const { isItemHovered, hoverProps } = useIsMenuItemHovered();
    return /* @__PURE__ */ jsxs(Fragment, { children: [
      /* @__PURE__ */ jsx(
        RadixItem,
        {
          ref: forwardedRef,
          ..._mergeProps(focusProps, hoverProps, {
            onFocus: () => slots.tooltip && !disabled && openTooltip(true),
            onBlur: () => slots.tooltip && !disabled && closeTooltip()
          }),
          "aria-label": ariaLabel,
          "aria-labelledby": ariaLabelledBy,
          "aria-describedby": ariaDescribedBy,
          "aria-details": ariaDetails,
          role: "menuitem",
          asChild: true,
          children: /* @__PURE__ */ jsxs(
            Flex,
            {
              ...remainingProps,
              className: clsx(
                textStyleCSS({ textStyle: "base-emphasized" }),
                !isItemHovered && focusClassName,
                styles.item({ disabled }),
                consumerClassName
              ),
              style: consumerStyle,
              as,
              ref: setTooltipTriggerRef,
              disabled,
              "aria-disabled": disabled,
              children: [
                slots.prefix,
                /* @__PURE__ */ jsx("div", { className: styles.itemChildren, children: slots.freeform }),
                slots.suffix
              ]
            }
          )
        }
      ),
      isOpen && /* @__PURE__ */ jsx(MenuItemTooltipWrapper, { triggerElement: tooltipTriggerRef, children: slots.tooltip })
    ] });
  }
);
Link.displayName = "Menu.Link";
export {
  Link
};
