// @dynatrace/strato-components/esm/navigation/menu/Menu.js
import { jsx, jsxs } from "react/jsx-runtime";
import { Root as RadixRoot } from "@radix-ui/react-dropdown-menu";
import {
  Children,
  cloneElement,
  isValidElement,
  useMemo
} from "react";
import { isElement } from "react-is";
import { COMPONENT_NAME } from "./constants.js";
import { Content } from "./Content.js";
import { MenuContext } from "./contexts/MenuContext.js";
import { Group } from "./Group.js";
import { useMenuAutoClose } from "./hooks/useMenuAutoClose.js";
import { Intent } from "./Intent.js";
import { Item } from "./Item.js";
import { ItemTooltip } from "./ItemTooltip.js";
import { Label } from "./Label.js";
import { Link } from "./Link.js";
import { Prefix } from "./Prefix.js";
import { Sub } from "./Sub.js";
import { SubContent } from "./SubContent.js";
import { SubTrigger } from "./SubTrigger.js";
import { Suffix } from "./Suffix.js";
import { Trigger } from "./Trigger.js";
import { useControllableState } from "../../core/hooks/useControllableState.js";
import { mergeProps as _mergeProps } from "../../core/utils/merge-props.js";
function getMenuSlots(children) {
  const slots = {
    trigger: null,
    content: [],
    freeform: []
  };
  Children.forEach(children, (child) => {
    if (isElement(child)) {
      switch (child.type) {
        case Trigger:
          slots.trigger = child;
          break;
        case Content:
          slots.content.push(child);
          break;
        default:
          slots.freeform.push(child);
      }
    } else {
      slots.freeform.push(child);
    }
  });
  return slots;
}
function registerTabHandlers(children, handleTabKeyPressed, dataDtComponent) {
  return Children.map(children, (child) => {
    return isValidElement(child) && child.type === Content ? cloneElement(
      child,
      _mergeProps(child.props, {
        _onTabKeyPressed: handleTabKeyPressed,
        "data-dt-component": dataDtComponent
      }),
      child.props.children
    ) : child;
  });
}
const MenuRoot = (props) => {
  const {
    open: value,
    onOpenChange: onChange,
    modal = true,
    defaultOpen = false,
    children,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    // @ts-expect-error - We want to allow overriding data-dt-component as we use the component internally in other Strato components, but it should not appear in the public API.
    "data-dt-component": dataDtComponent = COMPONENT_NAME
  } = props;
  const [open, updateOpen] = useControllableState({
    defaultValue: defaultOpen,
    value,
    onChange
  });
  const slots = useMemo(
    () => getMenuSlots(
      registerTabHandlers(children, () => updateOpen(false), dataDtComponent)
    ),
    [children, updateOpen, dataDtComponent]
  );
  useMenuAutoClose(open, updateOpen);
  const menuContextValue = useMemo(
    () => ({
      isOpen: open,
      isDefaultOpen: defaultOpen,
      setIsOpen: updateOpen
    }),
    [defaultOpen, open, updateOpen]
  );
  return /* @__PURE__ */ jsx(MenuContext.Provider, { value: menuContextValue, children: /* @__PURE__ */ jsxs(
    RadixRoot,
    {
      open,
      onOpenChange: (isOpen) => updateOpen(isOpen),
      modal,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      children: [
        slots.trigger !== null && slots.trigger,
        open && slots.content,
        slots.freeform
      ]
    }
  ) });
};
MenuRoot.displayName = "Menu";
const Menu = Object.assign(MenuRoot, {
  Trigger,
  Content,
  Item,
  ItemTooltip,
  Label,
  Link,
  Group,
  Prefix,
  Suffix,
  Sub,
  SubTrigger,
  SubContent,
  Intent
});
export {
  Menu
};
