// @dynatrace/strato-components/esm/navigation/menu/Sub.js
import { jsx, jsxs } from "react/jsx-runtime";
import { Sub as RadixSub } from "@radix-ui/react-dropdown-menu";
import {
  useMemo,
  Children,
  isValidElement,
  cloneElement
} from "react";
import { isElement } from "react-is";
import { MenuSubContext } from "./contexts/MenuSubContext.js";
import { useMenuAutoClose } from "./hooks/useMenuAutoClose.js";
import { SubContent } from "./SubContent.js";
import { SubTrigger } from "./SubTrigger.js";
import { useControllableState } from "../../core/hooks/useControllableState.js";
import { mergeProps as _mergeProps } from "../../core/utils/merge-props.js";
function registerTabHandlers(children, handleTabKeyPressed) {
  return Children.map(children, (child) => {
    return isValidElement(child) && child.type === SubContent ? cloneElement(
      child,
      _mergeProps(child.props, {
        _onTabKeyPressed: handleTabKeyPressed
      }),
      child.props.children
    ) : child;
  });
}
function getSubSlots(children) {
  const slots = {
    trigger: null,
    content: [],
    freeform: []
  };
  Children.forEach(children, (child) => {
    if (isElement(child)) {
      switch (child.type) {
        case SubTrigger:
          slots.trigger = child;
          break;
        case SubContent:
          slots.content.push(child);
          break;
        default:
          slots.freeform.push(child);
      }
    } else {
      slots.freeform.push(child);
    }
  });
  return slots;
}
const Sub = (props) => {
  const {
    defaultOpen: defaultValue,
    open: value,
    onOpenChange: onChange,
    children,
    ...remainingProps
  } = props;
  const [open, updateOpen] = useControllableState({
    defaultValue,
    value,
    onChange
  });
  const slots = useMemo(
    () => getSubSlots(registerTabHandlers(children, () => updateOpen(false))),
    [children, updateOpen]
  );
  useMenuAutoClose(open, updateOpen);
  return /* @__PURE__ */ jsx(
    MenuSubContext.Provider,
    {
      value: {
        isOpen: open ?? false,
        isDefaultOpen: defaultValue ?? false,
        setIsOpen: updateOpen
      },
      children: /* @__PURE__ */ jsxs(
        RadixSub,
        {
          open,
          onOpenChange: (isOpen) => updateOpen(isOpen),
          ...remainingProps,
          children: [
            slots.trigger !== null && slots.trigger,
            open && slots.content,
            slots.freeform
          ]
        }
      )
    }
  );
};
Sub.displayName = "Menu.Sub";
export {
  Sub
};
