// @dynatrace/strato-components/esm/navigation/menu/SubContent.js
import { jsx } from "react/jsx-runtime";
import {
  SubContent as RadixSubContent,
  Portal as RadixPortal
} from "@radix-ui/react-dropdown-menu";
import {
  forwardRef,
  useCallback
} from "react";
import * as styles from "./Menu.sty.js";
import { useOverlayStacking } from "../../core/hooks/useOverlayStacking.js";
const SUB_CONTENT_SIDE_OFFSET = 0;
const ALIGN_OFFSET = -4;
const SubContent = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const { _onTabKeyPressed, children } = props;
  const tabKeyHandler = useCallback(
    (e) => {
      if (_onTabKeyPressed && e.key === "Tab") {
        _onTabKeyPressed();
      }
    },
    [_onTabKeyPressed]
  );
  const { style } = useOverlayStacking(void 0, "overlay");
  return /* @__PURE__ */ jsx(RadixPortal, { children: /* @__PURE__ */ jsx(
    RadixSubContent,
    {
      ref: forwardedRef,
      sideOffset: SUB_CONTENT_SIDE_OFFSET,
      alignOffset: ALIGN_OFFSET,
      onKeyUp: tabKeyHandler,
      className: styles.content,
      style: { zIndex: style.zIndex },
      collisionPadding: 2,
      children
    }
  ) });
});
SubContent.displayName = "Menu.SubContent";
export {
  ALIGN_OFFSET,
  SubContent
};
