// @dynatrace/strato-components/esm/navigation/menu/SubTrigger.js
import { jsx, jsxs } from "react/jsx-runtime";
import { SubTrigger as RadixSubTrigger } from "@radix-ui/react-dropdown-menu";
import clsx from "clsx";
import { forwardRef } from "react";
import { ChevronRightSmallIcon } from "@dynatrace/strato-icons";
import { useIsMenuItemHovered } from "./hooks/useIsMenuItemHovered.js";
import { getSubTriggerSlots } from "./item-slots.js";
import * as styles from "./Menu.sty.js";
import { useFocusRing } from "../../core/styles/useFocusRing.js";
import { mergeProps as _mergeProps } from "../../core/utils/merge-props.js";
import { Flex } from "../../layouts/flex/Flex.js";
import { textStyle as textStyleCSS } from "../../styles/textStyle.sty.js";
const SubTrigger = /* @__PURE__ */ forwardRef(
  ({
    children,
    disabled,
    "aria-label": ariaLabel,
    "aria-labelledby": ariaLabelledBy,
    "aria-describedby": ariaDescribedBy,
    "aria-details": ariaDetails,
    ...remainingProps
  }, forwardedRef) => {
    const slots = getSubTriggerSlots(children);
    const { focusProps, focusClassName } = useFocusRing({
      disabled
    });
    const { isItemHovered, hoverProps } = useIsMenuItemHovered();
    return /* @__PURE__ */ jsx(
      RadixSubTrigger,
      {
        ref: forwardedRef,
        disabled,
        className: clsx(
          textStyleCSS({ textStyle: "base-emphasized" }),
          !isItemHovered && focusClassName,
          styles.item({ disabled })
        ),
        ..._mergeProps(remainingProps, focusProps, hoverProps),
        "aria-label": ariaLabel,
        "aria-labelledby": ariaLabelledBy,
        "aria-describedby": ariaDescribedBy,
        "aria-details": ariaDetails,
        "aria-haspopup": true,
        role: "menuitem",
        asChild: true,
        children: /* @__PURE__ */ jsxs(Flex, { justifyContent: "space-between", children: [
          slots.prefix,
          /* @__PURE__ */ jsx("div", { className: styles.subChildren, children: slots.freeform }),
          /* @__PURE__ */ jsx(ChevronRightSmallIcon, {})
        ] })
      }
    );
  }
);
SubTrigger.displayName = "Menu.SubTrigger";
export {
  SubTrigger
};
