// @dynatrace/strato-components/esm/filters/timeframe-selector/components/overlay/TimeframeSelectorOverlay.js
import { jsx, jsxs } from "react/jsx-runtime";
import { TZDateMini } from "@date-fns/tz";
import { forwardRef } from "react";
import { useIntl } from "react-intl";
import * as styles from "./TimeframeSelectorOverlay.sty.js";
import { TimeframeSelectorOverlayFooter } from "./TimeframeSelectorOverlayFooter.js";
import { FocusScope } from "../../../../core/components/focus-scope/FocusScope.js";
import { OverlayContainer } from "../../../../core/components/OverlayContainer.js";
import { useOverlayStacking } from "../../../../core/hooks/useOverlayStacking.js";
import { useRovingFocus } from "../../../../core/hooks/useRovingFocus.js";
import { OverlayStackingProvider } from "../../../../core/providers/OverlayStackingProvider.js";
import { useTimeframeSelectorStore } from "../../../../core/time/timeframeStore/useTimeframeStore.js";
import { useDateTimeFormat } from "../../../../forms/date-time-picker/hooks/useDateTimeFormat.js";
import { Flex } from "../../../../layouts/flex/Flex.js";
import { DismissButton } from "../../../../overlays/hooks/useOverlay.js";
import { TIMEFRAME_SELECTOR_MESSAGES } from "../../constants/timeframeSelectorMessages.js";
import { CustomTimeframeSelector } from "../custom-timeframe-selector/CustomTimeframeSelector.js";
import { Presets } from "../presets/Presets.js";
import { RecentTimeframes } from "../presets/RecentTimeframes.js";
function getMaxWidth(toParts) {
  const placeholderFromOrToParts = toParts(new TZDateMini()).map(
    (part) => {
      if (part.type === "dayPeriod") {
        return "AM/PM";
      }
      return part.value;
    }
  );
  const placeholderFromOrTo = placeholderFromOrToParts.join("");
  const placeholdersNumberOfChars = 2 * placeholderFromOrTo.length;
  const surroundingSpace = 22;
  return `${placeholdersNumberOfChars * 1.05 + // try to compensate for the ch unit being calculated for "0" instead of a placeholder char, which needs slightly more width
  surroundingSpace}ch`;
}
const TimeframeSelectorOverlay = forwardRef((props, forwardedRef) => {
  const {
    overlayProps,
    /* eslint-disable @typescript-eslint/no-deprecated */
    clearable,
    children,
    overlayClose,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow
  } = props;
  const intl = useIntl();
  const {
    trapFocus,
    style: overlayPropsStyle,
    ...remainingOverlayProps
  } = overlayProps;
  const overlayContainerProps = useOverlayStacking(void 0, "overlay");
  const precision = useTimeframeSelectorStore((state) => state.precision);
  const { toParts } = useDateTimeFormat("datetime", precision);
  const { containerRef } = useRovingFocus({
    initialIndex: 0,
    direction: "vertical",
    focusableChildrenSelector: "[role='option'][aria-disabled='false']"
  });
  return /* @__PURE__ */ jsx(
    OverlayContainer,
    {
      ...overlayContainerProps,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      children: /* @__PURE__ */ jsx(OverlayStackingProvider, { children: /* @__PURE__ */ jsxs(FocusScope, { autoFocus: true, children: [
        /* @__PURE__ */ jsxs(
          Flex,
          {
            ref: forwardedRef,
            gap: 4,
            flexDirection: "column",
            padding: 4,
            className: styles.wrapper,
            style: {
              ...overlayPropsStyle,
              width: getMaxWidth(toParts),
              maxHeight: `min(${overlayPropsStyle?.maxHeight}, 600px)`
            },
            "aria-label": intl.formatMessage(
              TIMEFRAME_SELECTOR_MESSAGES.overlayContainer
            ),
            ...remainingOverlayProps,
            children: [
              /* @__PURE__ */ jsx(CustomTimeframeSelector, {}),
              /* @__PURE__ */ jsxs(
                Flex,
                {
                  ref: containerRef,
                  gap: 4,
                  paddingLeft: 0,
                  paddingRight: 0,
                  flexDirection: "column",
                  className: styles.scrollContainer,
                  children: [
                    children ?? /* @__PURE__ */ jsx(Presets, {}),
                    /* @__PURE__ */ jsx(RecentTimeframes, {})
                  ]
                }
              ),
              /* @__PURE__ */ jsx(TimeframeSelectorOverlayFooter, { clearable })
            ]
          }
        ),
        /* @__PURE__ */ jsx(DismissButton, { onDismiss: overlayClose })
      ] }) })
    }
  );
});
TimeframeSelectorOverlay.displayName = "TimeframeSelectorOverlay";
export {
  TimeframeSelectorOverlay
};
