// @dynatrace/strato-components/esm/filters/timeframe-selector/components/overlay/TimeframeSelectorOverlayFooter.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { TZDateMini } from "@date-fns/tz";
import clsx from "clsx";
import { useMemo } from "react";
import { useIntl } from "react-intl";
import Spacings from "@dynatrace/strato-design-tokens/spacings";
import { getTimezone } from "@dynatrace-sdk/user-preferences";
import * as styles from "./TimeframeSelectorOverlay.sty.js";
import { Button } from "../../../../buttons/button/Button.js";
import { TIMEFRAME_SELECTOR_DOCUMENTATION_URL } from "../../../../core/external-links.js";
import { useTimeframeSelectorStore } from "../../../../core/time/timeframeStore/useTimeframeStore.js";
import { getSafeRegionalFormat } from "../../../../forms/forms-core/time/helpers.js";
import { Divider } from "../../../../layouts/divider/Divider.js";
import { Flex } from "../../../../layouts/flex/Flex.js";
import { globalSprinkles } from "../../../../styles/safe-sprinkles.js";
import { ExternalLink } from "../../../../typography/external-link/ExternalLink.js";
import { Text } from "../../../../typography/text/Text.js";
import { TIMEFRAME_SELECTOR_MESSAGES } from "../../constants/timeframeSelectorMessages.js";
const TimeframeSelectorOverlayFooter = (props) => {
  const { clearable } = props;
  const intl = useIntl();
  const setValue = useTimeframeSelectorStore(
    (state) => state.setTimeframeSelectorTimeframe
  );
  const timeZoneWithOffset = useMemo(() => {
    const timeZone = getTimezone();
    const parts = new Intl.DateTimeFormat(getSafeRegionalFormat(), {
      timeZone,
      timeZoneName: "shortOffset"
    }).formatToParts(TZDateMini.tz(timeZone));
    const partsShortName = new Intl.DateTimeFormat(getSafeRegionalFormat(), {
      timeZone,
      timeZoneName: "short"
    }).formatToParts(TZDateMini.tz(timeZone));
    const timeZonePart = parts?.find((part) => part.type === "timeZoneName")?.value.replace("GMT", "UTC");
    const timeZoneInfo = timeZonePart === "UTC" ? "UTC+0" : timeZonePart;
    const timezoneDetails = timeZoneInfo ? ` (${timeZoneInfo})` : "";
    const timeZoneShortName = partsShortName?.find(
      (part) => part.type === "timeZoneName"
    )?.value;
    return `${timeZoneShortName}${timezoneDetails}`;
  }, []);
  function handleClearSelection() {
    setValue(null);
  }
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(Divider, {}),
    /* @__PURE__ */ jsxs(
      Flex,
      {
        justifyContent: "space-between",
        alignItems: "center",
        flexWrap: "wrap-reverse",
        flex: 1,
        gap: 4,
        className: styles.footer,
        children: [
          /* @__PURE__ */ jsxs(
            Flex,
            {
              justifyContent: clearable ? "flex-start" : "space-between",
              alignItems: "center",
              flexShrink: 1,
              flexGrow: 1,
              className: clsx(globalSprinkles({ paddingLeft: 6 }), {
                [globalSprinkles({ paddingTop: 4, paddingBottom: 4 })]: !clearable
              }),
              children: [
                /* @__PURE__ */ jsx(Text, { className: styles.hint, children: timeZoneWithOffset }),
                clearable && /* @__PURE__ */ jsx(Divider, { orientation: "vertical" }),
                /* @__PURE__ */ jsx(
                  Text,
                  {
                    as: ExternalLink,
                    href: TIMEFRAME_SELECTOR_DOCUMENTATION_URL,
                    textStyle: "small",
                    style: { paddingRight: Spacings.Size6 },
                    "data-dt-name": "LearnMore",
                    children: intl.formatMessage(TIMEFRAME_SELECTOR_MESSAGES.docsLinkText)
                  }
                )
              ]
            }
          ),
          clearable && /* @__PURE__ */ jsx(
            Button,
            {
              type: "reset",
              onPointerDown: handleClearSelection,
              size: "condensed",
              children: intl.formatMessage(
                TIMEFRAME_SELECTOR_MESSAGES.resetTimeframeButton
              )
            }
          )
        ]
      }
    )
  ] });
};
TimeframeSelectorOverlayFooter.displayName = "TimeframeSelectorOverlayFooter";
export {
  TimeframeSelectorOverlayFooter
};
