// @dynatrace/strato-components/esm/filters/filter-field/utils/auto-format-logical-operators.js
import { ChangeSet } from "@codemirror/state";
function autoFormatLogicalOperators(value, tree, cursorPosition) {
  let newValue = value;
  const transactions = [];
  const logicalOperators = value.matchAll(/\b(and|or)\b/gi);
  for (const operator of logicalOperators) {
    const { index = -1 } = operator;
    if (index === -1) {
      continue;
    }
    if (tree.cursorAt(index, 1).name === "LogicalOperator") {
      const endIndex = index + operator[0].length;
      const capitalize = value.substring(endIndex, endIndex + 1) === " " || endIndex + 1 === cursorPosition && tree.cursorAt(endIndex + 1, 1).name === "LogicalOperator";
      const _operator = value.substring(index, endIndex);
      const newDoc = capitalize ? _operator.toUpperCase() : _operator;
      transactions.push({
        changes: ChangeSet.of(
          {
            from: index,
            to: endIndex,
            insert: newDoc
          },
          value.length
        ),
        selection: { anchor: cursorPosition },
        sequential: true
      });
      newValue = `${newValue.substring(0, index)}${capitalize ? _operator.toUpperCase() : _operator}${newValue.substring(endIndex)}`;
    }
  }
  return [newValue, transactions];
}
export {
  autoFormatLogicalOperators
};
