// @dynatrace/strato-components/esm/filters/filter-field/utils/get-next-suggestion.js
function clampIndex(index, length, wrapAround) {
  if (wrapAround) {
    if (index >= length) {
      return 0;
    } else if (index < 0) {
      return length - 1;
    }
  }
  return Math.max(0, Math.min(index, length - 1));
}
function getAllFocusableListItems(container) {
  if (!container) {
    return null;
  }
  return Array.from(
    container.querySelectorAll(
      "[data-strato-filter-field-virtual-focus-id]"
    )
  );
}
function getCurrentFocusableListItemIndex(container, currentSuggestionId) {
  if (!container || !currentSuggestionId) {
    return -1;
  }
  const currentSuggestion = getCurrentFocusableListItem(
    container,
    currentSuggestionId
  );
  const allFocusableListItems = getAllFocusableListItems(container);
  if (currentSuggestion && allFocusableListItems) {
    return allFocusableListItems.indexOf(currentSuggestion);
  }
  return -1;
}
function getCurrentFocusableListItemByIndex(container, index) {
  if (!container) {
    return null;
  }
  const allFocusableListItems = getAllFocusableListItems(container);
  if (allFocusableListItems && index >= 0 && index < allFocusableListItems.length) {
    return allFocusableListItems[index] ?? null;
  }
  return null;
}
function getCurrentFocusableListItem(container, currentSuggestionId) {
  if (!container || !currentSuggestionId || // Weird edge case handling as at least the JSDOM query selector will return
  // the entire parent if the selector contains a undefined or null string.
  currentSuggestionId === "null" || currentSuggestionId === "undefined") {
    return null;
  }
  return container.querySelector(
    `[data-strato-filter-field-virtual-focus-id="${currentSuggestionId}"]`
  );
}
function getFirstFocusableListItem(container) {
  if (!container) {
    return null;
  }
  return container.querySelector(
    "[data-strato-filter-field-virtual-focus-id]"
  );
}
function getLastFocusableListItem(container) {
  if (!container) {
    return null;
  }
  const allFocusableElements = Array.from(
    container.querySelectorAll(
      "[data-strato-filter-field-virtual-focus-id]"
    )
  );
  return Array.from(allFocusableElements).at(-1) ?? null;
}
function getNextFocusableListItem(direction, container, currentSuggestionId, nthSibling = 1) {
  if (!container) {
    return null;
  }
  const currentSuggestion = getCurrentFocusableListItem(
    container,
    currentSuggestionId
  );
  const allFocusableListItems = getAllFocusableListItems(container);
  if (currentSuggestion) {
    const currentIndex = allFocusableListItems?.indexOf(currentSuggestion);
    if (currentIndex !== void 0 && currentIndex > -1 && allFocusableListItems) {
      const modifier = direction === "next" ? nthSibling : nthSibling * -1;
      const newIndex = clampIndex(
        currentIndex + modifier,
        allFocusableListItems.length,
        nthSibling === 1
      );
      return allFocusableListItems[newIndex] ?? null;
    }
  } else if (direction === "next") {
    return allFocusableListItems?.at(
      // Count the nthSibling from the start depending on the step
      clampIndex(0 + (nthSibling - 1), allFocusableListItems.length, false)
    ) ?? null;
  } else if (direction === "previous") {
    return allFocusableListItems?.at(
      // Count the nthSibling from the end depending on the step
      clampIndex(nthSibling, allFocusableListItems.length, false) * -1
    ) ?? null;
  }
  return null;
}
export {
  getCurrentFocusableListItem,
  getCurrentFocusableListItemByIndex,
  getCurrentFocusableListItemIndex,
  getFirstFocusableListItem,
  getLastFocusableListItem,
  getNextFocusableListItem
};
