// @dynatrace/strato-components/esm/filters/filter-field/utils/migrate-validator-map.js
import {
  isFilterFieldValidatorMapDurationSpecification,
  isFilterFieldValidatorMapTypeSpecification,
  isFilterFieldValueSuggestionGroupConfig
} from "./validation-utils/type-guards.js";
function migrateValuePredicate(valuePredicate) {
  if (typeof valuePredicate === "function") {
    return { valuePredicate };
  }
  if (isFilterFieldValidatorMapTypeSpecification(valuePredicate)) {
    const raw = valuePredicate.type;
    const types = Array.isArray(raw) ? raw : [raw];
    const valueType = normalizeCollectedTypes([...new Set(types)]);
    return { valueType };
  }
  if (isFilterFieldValidatorMapDurationSpecification(valuePredicate)) {
    return { valuePredicate: `${valuePredicate.value}${valuePredicate.unit}` };
  }
  if (isFilterFieldValueSuggestionGroupConfig(valuePredicate)) {
    return {
      valuePredicate: {
        ...valuePredicate,
        suggestions: valuePredicate.suggestions.map(migrateValuePredicateItem)
      }
    };
  }
  if (Array.isArray(valuePredicate)) {
    return migrateValuePredicateArray(valuePredicate);
  }
  return { valuePredicate };
}
function migrateFilterFieldKeyPredicatesToArray(validatorMap) {
  const { keyPredicates } = validatorMap;
  if (keyPredicates === void 0 || Array.isArray(keyPredicates)) {
    return validatorMap;
  }
  const result = Object.entries(
    keyPredicates
  ).map(([key, config]) => {
    if (isFilterFieldValidatorMapTypeSpecification(config)) {
      const raw = config.type;
      const types = Array.isArray(raw) ? raw : [raw];
      return { key, valueType: normalizeCollectedTypes([...new Set(types)]) };
    }
    const { valuePredicate, ...rest } = config;
    if (valuePredicate === void 0) {
      return { key, ...rest };
    }
    const { valueType, valuePredicate: migratedPredicate } = migrateValuePredicate(valuePredicate);
    return {
      key,
      ...rest,
      ...valueType !== void 0 && { valueType },
      ...migratedPredicate !== void 0 && {
        valuePredicate: migratedPredicate
      }
    };
  });
  return { ...validatorMap, keyPredicates: result };
}
function migrateValuePredicateItem(item) {
  if (isFilterFieldValidatorMapDurationSpecification(item)) {
    return `${item.value}${item.unit}`;
  }
  return item;
}
function normalizeCollectedTypes(types) {
  if (types.length === 0) {
    return void 0;
  }
  return types.length === 1 ? types[0] : types;
}
function migrateValuePredicateArray(items) {
  const collectedTypes = [];
  const remaining = [];
  for (const item of items) {
    if (isFilterFieldValidatorMapTypeSpecification(item)) {
      const types = Array.isArray(item.type) ? item.type : [item.type];
      for (const t of types) {
        if (!collectedTypes.includes(t)) {
          collectedTypes.push(t);
        }
      }
    } else if (isFilterFieldValueSuggestionGroupConfig(item)) {
      remaining.push({
        ...item,
        suggestions: item.suggestions.map(migrateValuePredicateItem)
      });
    } else {
      remaining.push(migrateValuePredicateItem(item));
    }
  }
  const valueType = normalizeCollectedTypes(collectedTypes);
  const valuePredicate = remaining.length === 0 ? void 0 : remaining.length === 1 ? remaining[0] : remaining;
  return { valueType, valuePredicate };
}
export {
  migrateFilterFieldKeyPredicatesToArray
};
