// @dynatrace/strato-components/esm/filters/filter-field/utils/tree-node-guards.js
function isFilterFieldJSONPathNode(node) {
  return typeof node === "object" && node.type === "JSONPath";
}
function isFilterFieldStringNode(node) {
  return typeof node === "object" && node.type === "String";
}
function isFilterFieldVariableNode(node) {
  return typeof node === "object" && node.type === "Variable";
}
function isFilterFieldNumberNode(node) {
  return typeof node === "object" && node.type === "Number";
}
function isFilterFieldBigIntNode(node) {
  return typeof node === "object" && node.type === "BigInt";
}
function isFilterFieldBooleanNode(node) {
  return typeof node === "object" && node.type === "Boolean";
}
function isFilterFieldDurationNode(node) {
  return typeof node === "object" && node.type === "Duration";
}
function isFilterFieldExistsOperatorNode(node) {
  return typeof node === "object" && node.type === "ExistsOperator";
}
function isFilterFieldSearchOperatorNode(node) {
  return typeof node === "object" && node.type === "SearchOperator";
}
function isFilterFieldInclusionOperatorNode(node) {
  return typeof node === "object" && node.type === "InclusionOperator";
}
function isFilterFieldErrorNode(node) {
  return typeof node === "object" && node.type === "Error";
}
function isFilterFieldKeyNode(node) {
  return typeof node === "object" && node.type === "Key";
}
function isFilterFieldComparisonOperatorNode(node) {
  return typeof node === "object" && node.type === "ComparisonOperator";
}
function isFilterFieldGroupNode(node) {
  return typeof node === "object" && node.children !== void 0;
}
function isFilterFieldStatementNode(node) {
  return typeof node === "object" && node.type === "Statement";
}
function isFilterFieldListNode(node) {
  return typeof node === "object" && node.textValue !== void 0 && Array.isArray(node.value);
}
function isFilterFieldValueNode(node) {
  return typeof node === "object" && node.textValue !== void 0;
}
function isFilterFieldLeafNode(node) {
  return typeof node === "object" && node.textValue !== void 0 && !Array.isArray(node.value);
}
function isFilterValue(token) {
  const valueTypes = [
    "String",
    "Number",
    "Boolean",
    "Duration",
    "List",
    "Error"
  ];
  return valueTypes.includes(token.type);
}
function isValidPositionOperator(node) {
  return node.comparisonOperator?.value === "=" || node.comparisonOperator?.value === "!=" || node.comparisonOperator?.value === "ends-with" || node.comparisonOperator?.value === "not-ends-with";
}
function isLogicalOperator(node) {
  return node && (node.type === "LogicalOperator" || node.type === "Error" && (node.value === "OR" || node.value === "AND"));
}
function isFilterFieldSuggestionGroup(value) {
  return typeof value === "object" && (value.label !== void 0 || value.suggestions !== void 0);
}
function isFilterFieldSuggestion(value) {
  const suggestionValue = value;
  return typeof value === "object" && (suggestionValue.value !== void 0 || suggestionValue.displayValue !== void 0 || suggestionValue.insertionStrategy !== void 0);
}
export {
  isFilterFieldBigIntNode,
  isFilterFieldBooleanNode,
  isFilterFieldComparisonOperatorNode,
  isFilterFieldDurationNode,
  isFilterFieldErrorNode,
  isFilterFieldExistsOperatorNode,
  isFilterFieldGroupNode,
  isFilterFieldInclusionOperatorNode,
  isFilterFieldJSONPathNode,
  isFilterFieldKeyNode,
  isFilterFieldLeafNode,
  isFilterFieldListNode,
  isFilterFieldNumberNode,
  isFilterFieldSearchOperatorNode,
  isFilterFieldStatementNode,
  isFilterFieldStringNode,
  isFilterFieldSuggestion,
  isFilterFieldSuggestionGroup,
  isFilterFieldValueNode,
  isFilterFieldVariableNode,
  isFilterValue,
  isLogicalOperator,
  isValidPositionOperator
};
