// @dynatrace/strato-components/esm/filters/filter-field/utils/tree-navigation.js
import { syntaxTree } from "@codemirror/language";
import { getTextFromRange } from "./get-text.js";
function getStatementCursorAtPosition(cursorPosition, editorState) {
  const tree = syntaxTree(editorState);
  let cursor = tree.cursorAt(cursorPosition, -1);
  if (cursor.node.name === "Filter" || cursor.node.name === "LogicalOperator") {
    cursor = tree.cursorAt(cursorPosition, 1);
  }
  while (cursor.node.name !== "FilterStatement" && cursor.node.name !== "LogicalOperator" && cursor.parent()) {
  }
  return cursor;
}
function hasNextStatement(cursorPosition, editorState) {
  const cursor = getStatementCursorAtPosition(cursorPosition, editorState);
  let hasNext = cursor.nextSibling() && (cursor.node.name === "FilterStatement" || cursor.node.name === "LogicalOperator" || cursor.node.name === "FilterGroup") && cursor.from > cursorPosition;
  if (hasNext) {
    return true;
  }
  const { to } = cursor.node;
  if (cursor.parent()) {
    hasNext = to === cursor.node.to - 1;
  }
  return hasNext;
}
function isCursorInsideGroup(editorState, cursorPosition) {
  const tree = syntaxTree(editorState);
  const cursor = tree.cursorAt(cursorPosition, -1);
  while (cursor) {
    if (["OpeningBracket", "FilterGroup"].includes(cursor.node?.name)) {
      return true;
    }
    if (!cursor.parent()) {
      break;
    }
  }
  return false;
}
function hasNextSibling(cursorPosition, editorState, options = {}) {
  const { ignorePositionOperators } = options;
  const tree = syntaxTree(editorState);
  let cursor = tree.cursorAt(cursorPosition, 1);
  if (hasNextStatement(cursorPosition, editorState)) {
    return true;
  }
  if (cursor.node.name === "Filter" || cursor.node.name === "FilterStatement") {
    cursor = tree.cursorAt(cursorPosition, -1);
  }
  if (cursor.node.name === "OpeningBracket") {
    cursor.parent();
  }
  if (cursor.node.name === "ClosingBracket" && cursorPosition === cursor.node.from) {
    const bracketCursor = tree.cursorAt(cursor.node.from, -1);
    while (bracketCursor.prev() && bracketCursor.node.name !== "OpeningBracket") {
    }
    bracketCursor.prev();
    return bracketCursor.node.name === "FilterIncludesValue" || bracketCursor.node.name === "InclusionOperator";
  }
  let { to } = cursor.node;
  if (ignorePositionOperators && (cursor.node.name === "PositionOperator" || cursor.node.name === "InvalidPositionOperator")) {
    cursor.nextSibling();
    ({ to } = cursor.node);
  }
  if (cursor.node.name === "Identifier" || cursor.node.name === "DoubleQuotes" || cursor.node.name === "StringContent") {
    cursor.parent();
    ({ to } = cursor.node);
  }
  if (cursor.node.name === "String") {
    cursor.parent();
    ({ to } = cursor.node);
  }
  let nextSibling = cursor.nextSibling();
  while (nextSibling && cursor.node.name === "\u26A0") {
    nextSibling = cursor.nextSibling();
    ({ to } = cursor.node);
  }
  const hasNext = nextSibling && to < cursor.node.to;
  return hasNext;
}
function hasTrailingSpace(doc, node) {
  if (doc === void 0) {
    return false;
  }
  let text = doc;
  if (node) {
    text = getTextFromRange(node, doc.toString());
  }
  return text.slice(-1) === " ";
}
function getPositionOperators(node, tree) {
  let stringTreeCursor = tree.cursorAt(node.to, 1);
  const startsWithOperator = stringTreeCursor.node.name === "PositionOperator";
  stringTreeCursor = tree.cursorAt(node.from, -1);
  const endsWithOperator = stringTreeCursor.node.name === "PositionOperator";
  return { startsWithOperator, endsWithOperator };
}
function getPositionOperatorsInString(value) {
  const matches = value.matchAll(/(?:^\*?(?:"\*))|(?:(?:(?<!\\)\*")(?:\*?$))/g);
  let hasEndsWithOperator = false;
  let hasStartsWithOperator = false;
  for (const match of matches) {
    if (match.index === 0) {
      hasEndsWithOperator = true;
    }
    if (match.index > 0) {
      hasStartsWithOperator = true;
    }
  }
  return {
    hasEndsWithOperator,
    hasStartsWithOperator
  };
}
export {
  getPositionOperators,
  getPositionOperatorsInString,
  getStatementCursorAtPosition,
  hasNextSibling,
  hasNextStatement,
  hasTrailingSpace,
  isCursorInsideGroup
};
