// @dynatrace/strato-components/charts/core/components/tooltip/InteractiveChartTooltip/components/InteractiveChartTooltipVirtualizedSeries.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var InteractiveChartTooltipVirtualizedSeries_exports = {};
__export(InteractiveChartTooltipVirtualizedSeries_exports, {
  InteractiveChartTooltipVirtualizedSeries: () => InteractiveChartTooltipVirtualizedSeries
});
module.exports = __toCommonJS(InteractiveChartTooltipVirtualizedSeries_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_virtual = require("@tanstack/react-virtual");
var import_react = require("react");
var import_InteractiveChartTooltip_styles = require("./InteractiveChartTooltip.styles.js");
var import_Flex = require("../../../../../../layouts/flex/Flex.js");
var import_useHandleSetScrollbar = require("../hooks/useHandleSetScrollbar.js");
const InteractiveChartTooltipVirtualizedSeries = (props) => {
  const { itemSize, header, body, pointerEvents, onScrollContent } = props;
  const scrollRef = (0, import_react.useRef)(null);
  const scrollRefContainer = (0, import_react.useRef)(null);
  const { getVirtualItems, getTotalSize, measureElement } = (0, import_react_virtual.useVirtualizer)({
    count: body?.length ?? 0,
    getScrollElement: () => scrollRefContainer.current,
    estimateSize: (index) => {
      return itemSize?.(index) ?? 16;
    },
    overscan: 3
  });
  const items = getVirtualItems();
  (0, import_useHandleSetScrollbar.useHandleSetScrollbar)(getTotalSize());
  (0, import_react.useEffect)(() => {
    if (!onScrollContent) {
      return;
    }
    const scrollContainer = scrollRefContainer.current;
    scrollContainer?.addEventListener("scroll", onScrollContent);
    return () => {
      scrollContainer?.removeEventListener("scroll", onScrollContent);
    };
  }, [scrollRefContainer, onScrollContent]);
  const containerMinHeight = typeof itemSize === "undefined" ? Math.min((body?.length ?? 0) * 16, import_InteractiveChartTooltip_styles.TOOLTIP_MAX_HEIGHT) : Math.min((body?.length ?? 0) * itemSize(0), import_InteractiveChartTooltip_styles.TOOLTIP_MAX_HEIGHT);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_InteractiveChartTooltip_styles.StyledTooltipContainer,
    {
      minWidth: import_InteractiveChartTooltip_styles.DEFAULT_TOOLTIP_MIN_WIDTH,
      allowPointerEvents: pointerEvents,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { gap: 0, flexDirection: "column", children: [
        header,
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_InteractiveChartTooltip_styles.ScrollableContainerParent,
          {
            ref: scrollRefContainer,
            "data-testid": "chart-tooltip-item-container",
            style: {
              minHeight: `clamp(${containerMinHeight}px, ${getTotalSize()}px, ${import_InteractiveChartTooltip_styles.TOOLTIP_MAX_HEIGHT}px)`
            },
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_InteractiveChartTooltip_styles.TooltipItemsContainer,
              {
                ref: scrollRef,
                style: {
                  height: getTotalSize()
                },
                children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  "div",
                  {
                    style: {
                      transform: `translateY(${items[0]?.start ?? 0}px)`
                    },
                    children: items.map(({ index }) => {
                      const CurrentVirtualizedItem = body[index];
                      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                        import_InteractiveChartTooltip_styles.TooltipItems,
                        {
                          "data-testid": "chart-tooltip-item",
                          "data-index": index,
                          ref: measureElement,
                          children: CurrentVirtualizedItem()
                        },
                        index
                      );
                    })
                  }
                )
              }
            )
          }
        )
      ] })
    }
  );
};
InteractiveChartTooltipVirtualizedSeries["displayName"] = "InteractiveChartTooltipSeries";
