// @dynatrace/strato-components/charts/core/components/tooltip/InteractiveChartTooltip/components/InteractiveChartTooltip.sty.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var InteractiveChartTooltip_css_exports = {};
__export(InteractiveChartTooltip_css_exports, {
  formattedTimeframe: () => formattedTimeframe,
  scrollableContainerParent: () => scrollableContainerParent,
  titleContainer: () => titleContainer,
  tooltipHeader: () => tooltipHeader,
  tooltipItem: () => tooltipItem,
  tooltipItemsContainer: () => tooltipItemsContainer
});
module.exports = __toCommonJS(InteractiveChartTooltip_css_exports);
var import_textStyle_css_ts_vanilla = require("../../../../../../styles/textStyle.css");
var import_InteractiveChartTooltip_css_ts_vanilla = require("./InteractiveChartTooltip.css");
var formattedTimeframe = "_1fjte612-3-6-0";
var scrollableContainerParent = "_1fjte614-3-6-0";
var titleContainer = "_1fjte611-3-6-0";
var tooltipHeader = "_1fjte610-3-6-0 _da9a8v0-3-6-0 _da9a8v2-3-6-0 _da9a8v6-3-6-0 _da9a8ve-3-6-0";
var tooltipItem = "_1fjte615-3-6-0";
var tooltipItemsContainer = "_1fjte613-3-6-0";
