// @dynatrace/strato-components/charts/core/components/tooltip/InteractiveChartTooltip/components/InteractiveChartTooltip.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var InteractiveChartTooltip_exports = {};
__export(InteractiveChartTooltip_exports, {
  InteractiveChartTooltip: () => InteractiveChartTooltip
});
module.exports = __toCommonJS(InteractiveChartTooltip_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_BasicChartTooltip = require("./BasicChartTooltip/BasicChartTooltip.js");
var import_InteractiveChartTooltipVirtualizedSeries = require("./InteractiveChartTooltipVirtualizedSeries.js");
var import_useMergeRefs = require("../../../../../../core/hooks/useMergeRefs.js");
var import_is_slot = require("../../../../../../core/utils/is-slot.js");
var import_is_touchable_device = require("../../../../utils/is-touchable-device.js");
var import_useChartTooltip = require("../../shared/hooks/useChartTooltip.js");
var import_InteractiveTooltipContext = require("../context/InteractiveTooltipContext.js");
var import_Actions = require("../slots/Actions.js");
var import_Body = require("../slots/Body.js");
var import_Header = require("../slots/Header.js");
function iterateSlots(children) {
  const slots = { Header: void 0, Body: void 0, Actions: void 0 };
  import_react.Children.forEach(children, (child) => {
    if ((0, import_is_slot.isSlot)(child, import_Header.Header)) {
      slots.Header = child.props.children;
    } else if ((0, import_is_slot.isSlot)(child, import_Body.Body)) {
      slots.Body = child.props.children;
    } else if ((0, import_is_slot.isSlot)(child, import_Actions.Actions)) {
      slots.Actions = child.props.children;
    }
  });
  return slots;
}
const _InteractiveChartTooltip = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    pinned = false,
    chartRef,
    position,
    itemSize,
    updatePositionWhenPinned = false,
    children,
    onMouseLeave,
    onScrollContent
  } = props;
  const chartTooltipRef = (0, import_react.useRef)(null);
  const mergedRefs = (0, import_useMergeRefs.useMergeRefs)([forwardedRef, chartTooltipRef]);
  const [selectedItem, selectItem] = (0, import_react.useState)(
    void 0
  );
  const [hasScrollbar, setHasScrollbar] = (0, import_react.useState)(false);
  const onItemSelected = (e, item) => {
    e.stopPropagation();
    selectItem(item);
  };
  const { placement, updateTooltipPosition } = (0, import_useChartTooltip.useChartTooltip)(
    chartTooltipRef.current
  );
  (0, import_react.useEffect)(() => {
    if (!pinned || (0, import_is_touchable_device.isTouchableDevice)() || updatePositionWhenPinned) {
      updateTooltipPosition(position);
    }
  }, [
    pinned,
    position.x,
    position.y,
    updatePositionWhenPinned,
    updateTooltipPosition
  ]);
  (0, import_react.useEffect)(() => {
    return () => {
      onMouseLeave && onMouseLeave();
    };
  }, []);
  const { Header: Header2, Body: Body2 } = iterateSlots(children);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_InteractiveTooltipContext.InteractiveTooltipContext.Provider,
    {
      value: {
        selectedItem,
        selectItem: onItemSelected,
        hasScrollbar,
        setHasScrollbar
      },
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_BasicChartTooltip.BasicChartTooltip,
        {
          ref: mergedRefs,
          placement,
          chartRef,
          onMouseLeave,
          pinned,
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_InteractiveChartTooltipVirtualizedSeries.InteractiveChartTooltipVirtualizedSeries,
            {
              pinned,
              itemSize,
              header: Header2,
              body: Body2,
              pointerEvents: pinned,
              onScrollContent
            }
          )
        }
      )
    }
  );
});
_InteractiveChartTooltip.displayName = "InteractiveChartTooltip";
const InteractiveChartTooltip = Object.assign(_InteractiveChartTooltip, {
  Header: import_Header.Header,
  Body: import_Body.Body,
  Actions: import_Actions.Actions
});
