// @dynatrace/strato-components/charts/core/components/tooltip/InteractiveChartTooltip/components/InteractiveChartTooltip.styles.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var InteractiveChartTooltip_styles_exports = {};
__export(InteractiveChartTooltip_styles_exports, {
  DEFAULT_TOOLTIP_MIN_WIDTH: () => DEFAULT_TOOLTIP_MIN_WIDTH,
  ScrollableContainerParent: () => ScrollableContainerParent,
  StyledChartTooltipHeader: () => StyledChartTooltipHeader,
  StyledFormattedTimeframe: () => StyledFormattedTimeframe,
  StyledTitleContainer: () => StyledTitleContainer,
  StyledTooltipContainer: () => StyledTooltipContainer,
  TOOLTIP_MAX_HEIGHT: () => TOOLTIP_MAX_HEIGHT,
  TooltipItems: () => TooltipItems,
  TooltipItemsContainer: () => TooltipItemsContainer
});
module.exports = __toCommonJS(InteractiveChartTooltip_styles_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_spacings = __toESM(require("@dynatrace/strato-design-tokens/spacings"));
var styles = __toESM(require("./InteractiveChartTooltip.sty.js"));
var import_Flex = require("../../../../../../layouts/flex/Flex.js");
const TOOLTIP_MAX_HEIGHT = 200;
const TOOLTIP_MAX_WIDTH = 300;
const DEFAULT_TOOLTIP_MIN_WIDTH = "225px";
function StyledChartTooltipHeader(props) {
  const { className, ...remainingProps } = props;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      ...remainingProps,
      className: (0, import_clsx.default)(styles.tooltipHeader, className),
      tabIndex: -1
    }
  );
}
function StyledTitleContainer(props) {
  const { className, as = "div", ...remainingProps } = props;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Flex.Flex,
    {
      ...remainingProps,
      as,
      padding: 4,
      className: (0, import_clsx.default)(styles.titleContainer, className)
    }
  );
}
function StyledTooltipContainer(props) {
  const { style, minWidth, allowPointerEvents, ...remainingProps } = props;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      ...remainingProps,
      style: {
        ...style,
        minWidth,
        maxWidth: `calc(${TOOLTIP_MAX_WIDTH}px - (${import_spacings.default.Size8} * 2))`,
        pointerEvents: allowPointerEvents ? "auto" : "none",
        cursor: allowPointerEvents ? "pointer" : "auto"
      }
    }
  );
}
function StyledFormattedTimeframe(props) {
  const { className, ...remainingProps } = props;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "span",
    {
      ...remainingProps,
      className: (0, import_clsx.default)(styles.formattedTimeframe, className)
    }
  );
}
const TooltipItemsContainer = (0, import_react.forwardRef)((props, ref) => {
  const { className, ...remainingProps } = props;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      ref,
      ...remainingProps,
      className: (0, import_clsx.default)(styles.tooltipItemsContainer, className)
    }
  );
});
const ScrollableContainerParent = (0, import_react.forwardRef)((props, ref) => {
  const { className, style, ...remainingProps } = props;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      ref,
      ...remainingProps,
      className: (0, import_clsx.default)(styles.scrollableContainerParent, className),
      style: { ...style ?? {}, maxHeight: TOOLTIP_MAX_HEIGHT }
    }
  );
});
const TooltipItems = (0, import_react.forwardRef)(
  (props, ref) => {
    const { className, ...remainingProps } = props;
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "div",
      {
        ref,
        ...remainingProps,
        className: (0, import_clsx.default)(styles.tooltipItem, className)
      }
    );
  }
);
