// @dynatrace/strato-components/charts/core/components/tooltip/InteractiveChartTooltip/components/SeriesName.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SeriesName_exports = {};
__export(SeriesName_exports, {
  SeriesName: () => SeriesName
});
module.exports = __toCommonJS(SeriesName_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_OverlayStackingProvider = require("../../../../../../core/providers/OverlayStackingProvider.js");
var import_Tooltip = require("../../../../../../overlays/tooltip/Tooltip.js");
var import_constants = require("../../../../constants.js");
var import_useTruncationDetector = require("../../../../hooks/useTruncationDetector.js");
var import_single_line_name = require("../../../../utils/single-line-name.js");
var import_ChartTextEllipsis = require("../../../text-ellipsis/ChartTextEllipsis.js");
var import_SeriesBasedContent_css = require("../../shared/styles/SeriesBasedContent.sty.js");
const SeriesName = ({
  names,
  item,
  nameFormatter,
  singleLine,
  highlighted,
  truncationMode,
  isScrolling
}) => {
  const textRef = (0, import_react.useRef)(null);
  const name = nameFormatter?.(item) ?? (0, import_single_line_name.singleLineName)(item.name);
  const isTruncated = (0, import_useTruncationDetector.useTruncationDetector)(textRef, name);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayStackingProvider.OverlayStackingProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Tooltip.Tooltip, { text: name, disabled: isScrolling || !isTruncated, children: singleLine ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: (0, import_SeriesBasedContent_css.singleDimensionContainer)({ highlighted }), children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ChartTextEllipsis.ChartTextEllipsis,
    {
      ref: textRef,
      truncationMode: truncationMode ?? import_constants.DEFAULT_TRUNCATION_MODE,
      "aria-label": "Series name",
      children: name
    }
  ) }) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: (0, import_SeriesBasedContent_css.multipleDimensionContainer)({ highlighted }), children: names.map((dim, index) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "span",
    {
      className: import_SeriesBasedContent_css.multipleDimensionItem,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_ChartTextEllipsis.ChartTextEllipsis,
        {
          truncationMode: truncationMode ?? import_constants.DEFAULT_TRUNCATION_MODE,
          "aria-label": "Series name",
          children: dim
        }
      )
    },
    `dimension-${index}`
  )) }) }) });
};
SeriesName["displayName"] = "SeriesName";
