// @dynatrace/strato-components/esm/charts/core/components/data-point/utils/data-point.js
import { isEqual } from "lodash-es";
import Colors from "@dynatrace/strato-design-tokens/colors";
import { DEFAULT_DATA_POINT_RADIUS } from "../../../constants.js";
import { GROUP_PADDING } from "../constants.js";
const getStrokeColor = (currentPoints, closestPoint) => {
  if (!closestPoint) {
    return Colors.Text.Neutral.Default;
  }
  return currentPoints.find(
    (currentPoint) => isEqual(currentPoint, closestPoint)
  ) ? Colors.Text.Neutral.Default : Colors.Text.Neutral.Subdued;
};
function getOuterCircleSize(points, valueAxis) {
  const opCoordinateYX = valueAxis === "y" ? "x" : "y";
  const start = points[0][valueAxis];
  const end = points[points.length - 1][valueAxis];
  const minSize = Math.min(start, end);
  const maxSize = Math.max(start, end);
  const baseCoordinate = points[0][opCoordinateYX];
  const coordinateA = baseCoordinate - DEFAULT_DATA_POINT_RADIUS - GROUP_PADDING;
  const coordinateB = minSize - DEFAULT_DATA_POINT_RADIUS - GROUP_PADDING;
  const groupDiameterWithPadding = DEFAULT_DATA_POINT_RADIUS * 2 + GROUP_PADDING * 2;
  const size = maxSize - minSize + groupDiameterWithPadding;
  const diameter = groupDiameterWithPadding;
  if (valueAxis === "y") {
    return {
      x: coordinateA,
      y: coordinateB,
      width: diameter,
      height: size
    };
  }
  return {
    x: coordinateB,
    y: coordinateA,
    width: size,
    height: diameter
  };
}
export {
  getOuterCircleSize,
  getStrokeColor
};
