// @dynatrace/strato-geo/map/constants.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var constants_exports = {};
__export(constants_exports, {
  ACTIVE_BORDER_WIDTH_INNER: () => ACTIVE_BORDER_WIDTH_INNER,
  ACTIVE_BORDER_WIDTH_OUTER: () => ACTIVE_BORDER_WIDTH_OUTER,
  ACTIVE_COLOR: () => ACTIVE_COLOR,
  BASE_LAYER_FILL_ID: () => BASE_LAYER_FILL_ID,
  BASE_LAYER_IDS: () => BASE_LAYER_IDS,
  BASE_LAYER_LINE_ID: () => BASE_LAYER_LINE_ID,
  BOUNDING_BOX_PRECISION: () => BOUNDING_BOX_PRECISION,
  BUBBLE_DEFAULT_OPACITY: () => BUBBLE_DEFAULT_OPACITY,
  BUBBLE_OUTLINE_STROKE: () => BUBBLE_OUTLINE_STROKE,
  BUBBLE_STROKE: () => BUBBLE_STROKE,
  CDN_BASE_PATH: () => CDN_BASE_PATH,
  COPY_TO_CLIPBOARD_MESSAGES: () => COPY_TO_CLIPBOARD_MESSAGES,
  COUNTRY_BORDER_WIDTH: () => COUNTRY_BORDER_WIDTH,
  DEFAULT_BASE_LAYER_RULES: () => DEFAULT_BASE_LAYER_RULES,
  DEFAULT_BASIC_SHAPE_COLOR_TOKEN: () => DEFAULT_BASIC_SHAPE_COLOR_TOKEN,
  DEFAULT_BOUNDARIES_BORDER_COLOR: () => DEFAULT_BOUNDARIES_BORDER_COLOR,
  DEFAULT_BUBBLE_COLOR: () => DEFAULT_BUBBLE_COLOR,
  DEFAULT_CHOROPLETH_COLOR: () => DEFAULT_CHOROPLETH_COLOR,
  DEFAULT_COUNTRIES_FILL_COLOR: () => DEFAULT_COUNTRIES_FILL_COLOR,
  DEFAULT_COUNTRY_CODE: () => DEFAULT_COUNTRY_CODE,
  DEFAULT_FORMATTER: () => DEFAULT_FORMATTER,
  DEFAULT_ICON_BACKGROUND_COLOR: () => DEFAULT_ICON_BACKGROUND_COLOR,
  DEFAULT_ICON_COLOR: () => DEFAULT_ICON_COLOR,
  DEFAULT_ICON_SIZE_RATIO: () => DEFAULT_ICON_SIZE_RATIO,
  DEFAULT_INPUT_ICON_SIZE: () => DEFAULT_INPUT_ICON_SIZE,
  DEFAULT_LEGEND_SIZES: () => DEFAULT_LEGEND_SIZES,
  DEFAULT_LINE_CAP: () => DEFAULT_LINE_CAP,
  DEFAULT_LINE_COLOR: () => DEFAULT_LINE_COLOR,
  DEFAULT_LINE_DASH_ARRAY: () => DEFAULT_LINE_DASH_ARRAY,
  DEFAULT_LINE_JOIN: () => DEFAULT_LINE_JOIN,
  DEFAULT_LINE_THICKNESS: () => DEFAULT_LINE_THICKNESS,
  DEFAULT_MAP_HEIGHT: () => DEFAULT_MAP_HEIGHT,
  DEFAULT_MAP_LEGEND_RATIO: () => DEFAULT_MAP_LEGEND_RATIO,
  DEFAULT_MAP_POSITION: () => DEFAULT_MAP_POSITION,
  DEFAULT_RADIUS: () => DEFAULT_RADIUS,
  DEFAULT_RADIUS_EXPRESSION: () => DEFAULT_RADIUS_EXPRESSION,
  DEFAULT_RANGE_COLOR: () => DEFAULT_RANGE_COLOR,
  DEFAULT_SEQUENTIAL_LEGEND_COLOR_PALETTE: () => DEFAULT_SEQUENTIAL_LEGEND_COLOR_PALETTE,
  DEFAULT_SHAPE_COLOR: () => DEFAULT_SHAPE_COLOR,
  DEFAULT_SYMBOL_SIZE: () => DEFAULT_SYMBOL_SIZE,
  DEFAULT_TRUNCATION_MODE: () => DEFAULT_TRUNCATION_MODE,
  DIRECTION_ICON_OUTPUT_SIZE: () => DIRECTION_ICON_OUTPUT_SIZE,
  DIRECTION_ICON_SHRINK_RATIO: () => DIRECTION_ICON_SHRINK_RATIO,
  FALLBACK_SYMBOL_SIZE: () => FALLBACK_SYMBOL_SIZE,
  FIT_BOUNDS_OPTIONS: () => FIT_BOUNDS_OPTIONS,
  ICON_BACKGROUND_OPACITY: () => ICON_BACKGROUND_OPACITY,
  ICON_BACKGROUND_RADIUS: () => ICON_BACKGROUND_RADIUS,
  INCLUDE_ALL_WILDCARD_RULE: () => INCLUDE_ALL_WILDCARD_RULE,
  MAP_VIEW_ARIA_LABEL: () => MAP_VIEW_ARIA_LABEL,
  MAX_BOUNDS_PRECISION: () => MAX_BOUNDS_PRECISION,
  MAX_LATITUDE: () => MAX_LATITUDE,
  MAX_LINE_THICKNESS: () => MAX_LINE_THICKNESS,
  MAX_LONGITUDE: () => MAX_LONGITUDE,
  MIN_LATITUDE: () => MIN_LATITUDE,
  MIN_LINE_THICKNESS: () => MIN_LINE_THICKNESS,
  MIN_LONGITUDE: () => MIN_LONGITUDE,
  REGION_BORDER_WIDTH: () => REGION_BORDER_WIDTH,
  SHAPE_OPACITY_DEFAULT: () => SHAPE_OPACITY_DEFAULT,
  SHAPE_OPACITY_DIMMED: () => SHAPE_OPACITY_DIMMED,
  TRANSPARENT_BORDER_COLOR: () => TRANSPARENT_BORDER_COLOR,
  VIEW_STATE_PRECISION: () => VIEW_STATE_PRECISION,
  WORLD_VIEW_MAX_BOUNDARIES: () => WORLD_VIEW_MAX_BOUNDARIES
});
module.exports = __toCommonJS(constants_exports);
var import_react_intl = require("react-intl");
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
const DEFAULT_COUNTRIES_FILL_COLOR = import_colors.default.Background.Container.Neutral.Default;
const DEFAULT_BOUNDARIES_BORDER_COLOR = import_colors.default.Border.Neutral.Default;
const COUNTRY_BORDER_WIDTH = 1;
const REGION_BORDER_WIDTH = 0.4;
const TRANSPARENT_BORDER_COLOR = "#00000000";
const MAX_LATITUDE = 85;
const MIN_LATITUDE = -60;
const MAX_LONGITUDE = 179.99;
const MIN_LONGITUDE = -179.99;
const WORLD_VIEW_MAX_BOUNDARIES = [
  MAX_LONGITUDE,
  MAX_LATITUDE,
  MIN_LONGITUDE,
  MIN_LATITUDE
];
const DEFAULT_COUNTRY_CODE = "default";
const FIT_BOUNDS_OPTIONS = { padding: 20 };
const DEFAULT_MAP_HEIGHT = 400;
const DEFAULT_MAP_LEGEND_RATIO = "auto";
const DEFAULT_MAP_POSITION = "auto";
const DEFAULT_INPUT_ICON_SIZE = 20;
const ACTIVE_COLOR = import_colors.default.Border.Neutral.Accent;
const BASE_LAYER_FILL_ID = "geojson-fill";
const BASE_LAYER_LINE_ID = "geojson-line";
const BASE_LAYER_IDS = [BASE_LAYER_LINE_ID, BASE_LAYER_FILL_ID];
const SHAPE_OPACITY_DEFAULT = 1;
const SHAPE_OPACITY_DIMMED = 0.2;
const DEFAULT_SYMBOL_SIZE = 32;
const FALLBACK_SYMBOL_SIZE = 1;
const DEFAULT_TRUNCATION_MODE = "middle";
const CDN_BASE_PATH = "https://dt-cdn.net/scripts/data/worldmap/maps/v006/";
const INCLUDE_ALL_WILDCARD_RULE = "*";
const DEFAULT_BASIC_SHAPE_COLOR_TOKEN = import_colors.default.Charts.CategoricalThemed.PurpleRain.Color01;
const DEFAULT_LEGEND_SIZES = {
  // Minimum size in which the legend can be resized (vertical layout)
  minWidth: 0.1,
  // Minimum size in which the legend can be resized (horizontal layout)
  minHeight: 0.1,
  // Maximum size in which the legend can be resized (vertical layout). 80% represent 4 / 5ths of the container.
  maxWidth: 0.8,
  // Maximum size in which the legend can be resized (horizontal layout). 80% represent 4 / 5ths of the container.
  maxHeight: 0.8
};
const DEFAULT_FORMATTER = {
  maximumFractionDigits: 1
};
const DEFAULT_RANGE_COLOR = import_colors.default.Background.Container.Neutral.Accent;
const DEFAULT_SEQUENTIAL_LEGEND_COLOR_PALETTE = "blue";
const ICON_BACKGROUND_OPACITY = 0.7;
const ICON_BACKGROUND_RADIUS = 16;
const DEFAULT_ICON_BACKGROUND_COLOR = import_colors.default.Background.Container.Neutral.Subdued;
const DEFAULT_ICON_SIZE_RATIO = 0.6;
const DEFAULT_SHAPE_COLOR = import_colors.default.Charts.Categorical.Color15.Default;
const DEFAULT_ICON_COLOR = import_colors.default.Icon.Neutral.Default;
const DEFAULT_BUBBLE_COLOR = import_colors.default.Charts.Categorical.Color07.Default;
const DEFAULT_RADIUS = 12;
const BUBBLE_DEFAULT_OPACITY = 0.4;
const BUBBLE_STROKE = 1;
const BUBBLE_OUTLINE_STROKE = 1;
const DEFAULT_RADIUS_EXPRESSION = [
  "get",
  "__radius"
];
const DEFAULT_LINE_JOIN = "miter";
const DEFAULT_LINE_CAP = "butt";
const DEFAULT_LINE_THICKNESS = 2;
const DEFAULT_LINE_COLOR = import_colors.default.Charts.Categorical.Color03.Default;
const DEFAULT_LINE_DASH_ARRAY = [5, 2.5];
const MIN_LINE_THICKNESS = 1;
const MAX_LINE_THICKNESS = 64;
const DIRECTION_ICON_OUTPUT_SIZE = 128;
const DIRECTION_ICON_SHRINK_RATIO = 1 / 15;
const ACTIVE_BORDER_WIDTH_OUTER = 3;
const ACTIVE_BORDER_WIDTH_INNER = 1;
const DEFAULT_CHOROPLETH_COLOR = import_colors.default.Charts.Sequential.Blue.Color05.Default;
const COPY_TO_CLIPBOARD_MESSAGES = (0, import_react_intl.defineMessages)({
  copyActionCoords: {
    id: "g2qESM+ibB5sUyjf",
    defaultMessage: "Copy coordinates",
    description: "Text in a tooltip shown when hovering a tooltip item and hovering the copy button next to its coordinates"
  }
});
const DEFAULT_BASE_LAYER_RULES = {
  include: [INCLUDE_ALL_WILDCARD_RULE],
  exclude: []
};
const BOUNDING_BOX_PRECISION = 6;
const VIEW_STATE_PRECISION = 4;
const MAX_BOUNDS_PRECISION = 2;
const MAP_VIEW_ARIA_LABEL = "Map View";
