// @dynatrace/strato-geo/map/MapView.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MapView_exports = {};
__export(MapView_exports, {
  MapView: () => MapView
});
module.exports = __toCommonJS(MapView_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_charts = require("@dynatrace/strato-components/charts");
var import_core = require("@dynatrace/strato-components/core");
var import_MapLegendRenderer = require("./components/legend/MapLegendRenderer.js");
var import_MapContent = require("./components/MapContent.js");
var import_MapUnavailable = require("./components/MapUnavailable.js");
var import_constants = require("./constants.js");
var import_formatter_context = require("./contexts/formatter.context.js");
var import_layer_ids_context = require("./contexts/layer-ids.context.js");
var import_map_base_layer_context = require("./contexts/map-base-layer.context.js");
var import_map_configuration_context = require("./contexts/map-configuration.context.js");
var import_map_loading_context = require("./contexts/map-loading.context.js");
var import_map_raw_data_context = require("./contexts/map-raw-data.context.js");
var import_map_truncation_mode_context = require("./contexts/map-truncation-mode.context.js");
var import_use_load_map_base_layer = require("./hooks/use-load-map-base-layer.js");
var import_color_scale_provider = require("./providers/color-scale.provider.js");
var import_layer_color_strategy_provider = require("./providers/layer-color-strategy.provider.js");
var import_DownloadCSV = require("./slots/DownloadCSV.js");
var import_ErrorStateSlot = require("./slots/states/ErrorStateSlot.js");
var import_Toolbar = require("./slots/Toolbar.js");
var import_Zoom = require("./slots/Zoom.js");
var import_map_store_provider = require("./store/map-store.provider.js");
var import_build_scale_from_legend_config = require("./utils/build-scale-from-legend-config.js");
var import_extract_layers_data = require("./utils/extract-layers-data.js");
var import_get_map_states_template = require("./utils/get-map-states-template.js");
var import_iterate_config_slots = require("./utils/iterate-config-slots.js");
var import_set_layers_id = require("./utils/set-layers-id.js");
var import_react_mapgl_styles = require("./styles/react-mapgl-styles.css");
const _MapView = (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const [behavioralTrackingProps] = (0, import_core._useBehavioralTrackingProps)(props);
    const {
      children,
      loading = false,
      style: costumerStyle,
      className: customerClassNames,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "data-testid": dataTestId,
      formatter,
      truncationMode,
      ...remaining
    } = props;
    const { layerIds, parsedChildren, valueAccessors } = (0, import_set_layers_id.setLayersId)(children);
    const {
      flattenData: layersData,
      categories,
      legendDomain
    } = (0, import_extract_layers_data.extractLayersData)(parsedChildren, valueAccessors);
    const config = (0, import_iterate_config_slots.iterateConfigSlots)(children, legendDomain);
    const getMaxRange = () => {
      if (config.legend && (0, import_build_scale_from_legend_config.isThresholdLegend)(config.legend)) {
        return config.legend.ranges.at(-1)?.to ?? legendDomain[1];
      }
      return legendDomain[1];
    };
    const dataMax = getMaxRange();
    const isLegendHidden = !config.legend || !!config.legend.hidden;
    const legendPosition = config.legend?.position;
    const legendRatio = config.legend?.ratio;
    const legendOnRatioChange = config.legend?.onRatioChange;
    const { isMapEnabled, isFetchingFeatures, baseFeatureCollection, error } = (0, import_use_load_map_base_layer.useLoadMapBaseLayer)(config.baseLayer);
    const [isMapLoaded, setIsMapLoaded] = (0, import_react.useState)(false);
    const { errorState } = (0, import_get_map_states_template.getMapStatesTemplates)(children);
    const chartHeight = (0, import_charts._coerceSizeValue)(props.height) || import_constants.DEFAULT_MAP_HEIGHT;
    const containerStyles = { width: "100%", ...costumerStyle };
    const chartLayoutRef = (0, import_charts._useAutoLegendRefresh)(categories.join(":"));
    const mapUnavailable = !isFetchingFeatures && !isMapEnabled || error;
    const isLoading = isFetchingFeatures || !isMapLoaded;
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "div",
      {
        style: containerStyles,
        className: customerClassNames,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        "data-testid": dataTestId,
        role: "img",
        "aria-label": import_constants.MAP_VIEW_ARIA_LABEL,
        ...behavioralTrackingProps,
        // TODO: Error boundaries global fix needed - this error state only purpose is to make map not rendering when there error in response
        children: mapUnavailable ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_MapUnavailable.MapUnavailable,
          {
            isFetchingFeatures,
            isDisabled: !isMapEnabled,
            chartHeight,
            error
          }
        ) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_map_loading_context.MapLoadingContext.Provider, { value: loading, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_map_base_layer_context.MapBaseLayerFeaturesContext.Provider,
          {
            value: baseFeatureCollection,
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_map_configuration_context.MapConfigurationContext.Provider, { value: config, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_layer_ids_context.LayerIdsContext.Provider, { value: layerIds, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_map_truncation_mode_context.MapTruncationModeContext.Provider,
              {
                value: truncationMode ?? import_constants.DEFAULT_TRUNCATION_MODE,
                children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_formatter_context.FormatterContext.Provider, { value: formatter, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_map_raw_data_context.MapRawDataContext.Provider, { value: layersData, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_charts._OverlayProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_charts._OverlayTooltipProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_map_store_provider.MapStoreProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  import_color_scale_provider.ColorScaleProvider,
                  {
                    categories,
                    dataMax,
                    children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_layer_color_strategy_provider.LayerColorStrategyProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
                      import_charts._ChartLayout,
                      {
                        ref: chartLayoutRef,
                        chartHeight,
                        errorState,
                        showLoader: loading || isLoading,
                        children: [
                          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_charts._ChartLayout.Graph, { children: baseFeatureCollection && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                            import_MapContent.MapContent,
                            {
                              ref: forwardedRef,
                              onMapLoad: () => setIsMapLoaded(true),
                              ...remaining,
                              children: parsedChildren
                            }
                          ) }),
                          !isLegendHidden && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                            import_charts._ChartLayout.Legend,
                            {
                              position: legendPosition,
                              ratio: legendRatio,
                              ratioBoundaries: import_constants.DEFAULT_LEGEND_SIZES,
                              onResize: legendOnRatioChange,
                              children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_MapLegendRenderer.MapLegendRenderer, {})
                            }
                          )
                        ]
                      }
                    ) })
                  }
                ) }) }) }) }) })
              }
            ) }) })
          }
        ) })
      }
    );
  }
);
_MapView["displayName"] = "MapView";
const MapView = Object.assign(_MapView, {
  ErrorState: import_ErrorStateSlot.ErrorStateSlot,
  Toolbar: import_Toolbar.Toolbar,
  DownloadCSV: import_DownloadCSV.DownloadCSV,
  Zoom: import_Zoom.Zoom
});
