// @dynatrace/strato-components/esm/forms/calendar/Calendar.js
import { jsx } from "react/jsx-runtime";
import {
  forwardRef,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState
} from "react";
import { CalendarRoot } from "./CalendarRoot.js";
import {
  convertToCalendarStructure,
  getTimeOffsetFromInitialValue
} from "./utils/helpers.js";
import { useCallbackRef } from "../../core/hooks/useCallbackRef.js";
import { useMergeRefs } from "../../core/hooks/useMergeRefs.js";
import { isFullTimeframe } from "../../core/time/helpers.js";
import { getCalendarActions } from "../../core/time/timeframeStore/actions/calendar.js";
import { createTimeframeStore } from "../../core/time/timeframeStore/createTimeframeStore.js";
import { CalendarStoreProvider } from "../../core/time/timeframeStore/timeframeStoreProvider.js";
import { useMinMax, useMinMaxStoreSync } from "../../core/time/useMinMax.js";
import { validate } from "../../core/time/validation.js";
import { useFormFieldMessagesContext } from "../forms-core/hooks/useFormFieldMessagesContext.js";
import { useSyncedFormControlProps } from "../forms-core/hooks/useSyncedFormControlProps.js";
import { CALENDAR_ERROR_MESSAGES } from "./calendarMessages.js";
function useSyncPropsAndState(store, memoedValue, isControlled, type, precision, min, max, formFieldMessageDispatch, calendarRef) {
  const valueMode = type === "single" ? "single" : "timeframeSelector";
  useEffect(
    () => {
      isControlled && store.setState((state) => {
        const error = validate(memoedValue, state.min, state.max, {
          formMessageDispatch: formFieldMessageDispatch,
          formerError: state.calendarError,
          precision: state.precision,
          showError: false,
          checkOrder: true,
          ref: calendarRef,
          valueMode,
          errorMessages: CALENDAR_ERROR_MESSAGES
        });
        const timeOffset = getTimeOffsetFromInitialValue(
          memoedValue,
          type,
          min,
          max,
          precision
        );
        const valueWithPrecision = convertToCalendarStructure(
          type,
          memoedValue,
          min,
          max,
          precision
        );
        if (type === "single") {
          return {
            calendarValue: valueWithPrecision,
            calendarError: error,
            timeOffset
          };
        }
        const value = isFullTimeframe(valueWithPrecision) ? valueWithPrecision : null;
        return {
          calendarValue: value,
          calendarError: error,
          timeOffset
        };
      });
    },
    // calendarRef is a stable ref — no need to include it in deps
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [
      formFieldMessageDispatch,
      isControlled,
      memoedValue,
      store,
      min,
      max,
      precision,
      type
    ]
  );
  useEffect(() => {
    store.setState((state) => {
      const value = convertToCalendarStructure(
        type,
        state.calendarValue,
        min,
        max,
        precision
      );
      return {
        ...state,
        precision,
        calendarValue: value
      };
    });
  }, [precision, store, min, max, type]);
}
function RefAwareCalendar(props, forwardedRef) {
  const {
    value: propValue,
    defaultValue: propDefaultValue,
    onChange,
    min: propMin,
    max: propMax,
    disabled: propDisabled,
    required: propRequired,
    readOnly: propReadOnly,
    id: propId,
    "aria-disabled": propAriaDisabled,
    precision = "minutes",
    type = "single",
    ...remainingProps
  } = props;
  const [isControlled] = useState(propValue !== void 0);
  const onChangeProp = useCallbackRef(onChange);
  const { min, max } = useMinMax(propMin, propMax);
  const memoedValue = useMemo(() => {
    return convertToCalendarStructure(type, propValue, min, max, precision);
  }, [precision, propValue, min, max, type]);
  const memoedDefaultValue = useMemo(
    () => convertToCalendarStructure(type, propDefaultValue, min, max, precision),
    /** We're ignoring updates to defaultValue as it only considers the initial state. */
    // eslint-disable-next-line react-hooks/exhaustive-deps
    []
  );
  const calendarRef = useRef(null);
  const { dispatch: formFieldMessageDispatch } = useFormFieldMessagesContext();
  const { disabled, required, id, ariaDisabled, readOnly } = useSyncedFormControlProps({
    propDisabled,
    propRequired,
    propId,
    propAriaDisabled,
    propReadOnly
  });
  const [store] = useState(() => {
    const value = isControlled ? memoedValue : memoedDefaultValue;
    const error = validate(value, min, max, {
      formMessageDispatch: formFieldMessageDispatch,
      precision,
      showError: false,
      checkOrder: true,
      ref: calendarRef,
      valueMode: type === "single" ? "single" : "timeframeSelector",
      required,
      errorMessages: CALENDAR_ERROR_MESSAGES
    });
    const timeOffset = getTimeOffsetFromInitialValue(
      value,
      type,
      min,
      max,
      precision
    );
    const initialState = {
      calendarValue: value,
      min,
      max,
      precision,
      isFullDay: false,
      calendarRef,
      required,
      calendarError: error,
      timeOffset
    };
    const actions = getCalendarActions(
      isControlled,
      type,
      formFieldMessageDispatch,
      onChangeProp
    );
    return createTimeframeStore(initialState, actions);
  });
  useSyncPropsAndState(
    store,
    memoedValue,
    isControlled,
    type,
    precision,
    min,
    max,
    formFieldMessageDispatch,
    calendarRef
  );
  useMinMaxStoreSync(
    store,
    min,
    max,
    useCallback(
      (state, min2, max2) => ({
        calendarError: validate(state.calendarValue, min2, max2, {
          formMessageDispatch: formFieldMessageDispatch,
          precision,
          showError: false,
          checkOrder: true,
          ref: calendarRef,
          valueMode: type === "single" ? "single" : "timeframeSelector",
          required,
          errorMessages: CALENDAR_ERROR_MESSAGES
        })
      }),
      [formFieldMessageDispatch]
    )
  );
  const mergedRef = useMergeRefs([calendarRef, forwardedRef]);
  return /* @__PURE__ */ jsx(CalendarStoreProvider, { store, children: /* @__PURE__ */ jsx(
    CalendarRoot,
    {
      ...remainingProps,
      id,
      ref: mergedRef,
      "aria-disabled": ariaDisabled,
      disabled,
      required,
      readOnly,
      type
    }
  ) });
}
const Calendar = forwardRef(RefAwareCalendar);
export {
  Calendar
};
