// @dynatrace/strato-components/esm/forms/calendar/CalendarRoot.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import {
  forwardRef,
  useCallback,
  useImperativeHandle,
  useMemo,
  useRef
} from "react";
import { CalendarHeader } from "./calendar-header/CalendarHeader.js";
import { CALENDAR_ERROR_MESSAGES } from "./calendarMessages.js";
import { VisuallyHidden } from "../../core/components/visually-hidden/VisuallyHidden.js";
import { useCalendarStore } from "../../core/time/timeframeStore/useTimeframeStore.js";
import { validate } from "../../core/time/validation.js";
import { useFormFieldMessagesContext } from "../forms-core/hooks/useFormFieldMessagesContext.js";
import { useFormFieldValidation } from "../forms-core/hooks/useFormFieldValidation.js";
import { FormFieldMessagesAction } from "../forms-core/state/formfield-messages-state-reducer.js";
import * as styles from "./Calendar.sty.js";
import { CalendarContext } from "./contexts/CalendarContext.js";
import { useCalendar } from "./hooks/useCalendar.js";
import { useCalendarNavigation } from "./hooks/useCalendarNavigation.js";
import { SelectDayView } from "./views/SelectDayView.js";
import { SelectMonthView } from "./views/SelectMonthView.js";
import { SelectYearView } from "./views/SelectYearView.js";
import { isFullTimeframe, isTimeValue } from "../../core/time/helpers.js";
const COMPONENT_NAME = "Calendar";
function renderView(view) {
  switch (view) {
    case "month":
      return /* @__PURE__ */ jsx(SelectMonthView, {});
    case "year":
      return /* @__PURE__ */ jsx(SelectYearView, {});
    default:
      return /* @__PURE__ */ jsx(SelectDayView, {});
  }
}
function CalendarComponent(props, forwardedRef) {
  const {
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    id,
    disabled,
    "aria-disabled": ariaDisabled,
    required,
    name,
    readOnly = false,
    // eslint-disable-next-line @typescript-eslint/prefer-ts-expect-error, @typescript-eslint/ban-ts-comment
    // @ts-ignore Strict mode transition - needs to be investigated
    type = "single",
    "aria-label": ariaLabelProp,
    "aria-labelledby": ariaLabelledByProp,
    "aria-describedby": ariaDescribedByProp,
    "aria-details": ariaDetailsProp,
    ...remainingProps
  } = props;
  const calendarValue = useCalendarStore((state2) => state2.calendarValue);
  const min = useCalendarStore((state2) => state2.min);
  const max = useCalendarStore((state2) => state2.max);
  const error = useCalendarStore((state2) => state2.calendarError);
  const precision = useCalendarStore((state2) => state2.precision);
  const calendarRef = useRef(null);
  const ref = useRef(null);
  const refTo = useRef(null);
  const { uniqueId, state, dispatch } = useCalendar({
    type
  });
  useCalendarNavigation({
    calendarRef,
    view: state.currentView,
    dispatch
  });
  const { dispatch: formMessageDispatch } = useFormFieldMessagesContext();
  const refValidate = useCallback(() => {
    formMessageDispatch?.({
      type: FormFieldMessagesAction.SETISTOUCHED,
      payload: { isTouched: true }
    });
    validate(calendarValue, min, max, {
      formMessageDispatch,
      formerError: error,
      precision,
      showError: true,
      checkOrder: false,
      ref,
      errorMessages: CALENDAR_ERROR_MESSAGES
    });
  }, [min, max, formMessageDispatch, error, precision, ref, calendarValue]);
  useImperativeHandle(
    forwardedRef,
    () => {
      if (type === "single") {
        return {
          element: calendarRef.current,
          inputRef: ref.current,
          validate: refValidate
        };
      }
      return {
        element: calendarRef.current,
        inputRef: {
          from: ref?.current,
          to: refTo?.current
        },
        validate: refValidate
      };
    },
    [refValidate, type]
  );
  const {
    ariaProps,
    onChange: validationOnChange,
    onBlur,
    onInvalid
  } = useFormFieldValidation(
    ref,
    {},
    {
      "aria-label": ariaLabelProp,
      "aria-labelledby": ariaLabelledByProp,
      "aria-describedby": ariaDescribedByProp,
      "aria-details": ariaDetailsProp
    }
  );
  const labelRef = useRef(`calendar-body-label-${uniqueId}`);
  const disabledOrAriaDisabled = disabled || ariaDisabled || false;
  const getNativeProps = (suffix = "") => ({
    readOnly,
    required,
    disabled,
    "aria-disabled": ariaDisabled,
    name: name && `${name}${suffix}`,
    id
  });
  const calenderContextValue = useMemo(
    () => ({ state, dispatch, readOnly }),
    [state, dispatch, readOnly]
  );
  return /* @__PURE__ */ jsxs(CalendarContext.Provider, { value: calenderContextValue, children: [
    /* @__PURE__ */ jsxs(
      "div",
      {
        className: clsx(styles.calendar({ readOnly }), consumerClassName, {
          [styles.disabled]: disabledOrAriaDisabled
        }),
        "data-testid": dataTestId,
        ref: calendarRef,
        style: consumerStyle,
        ...remainingProps,
        "data-dt-component": COMPONENT_NAME,
        children: [
          /* @__PURE__ */ jsx(CalendarHeader, { view: state.currentView, labelId: labelRef.current }),
          /* @__PURE__ */ jsx("div", { "aria-labelledby": labelRef.current, children: /* @__PURE__ */ jsx(
            "table",
            {
              role: "grid",
              className: styles.body({
                isReadOnly: !!readOnly
              }),
              children: renderView(state.currentView)
            }
          ) })
        ]
      }
    ),
    /* @__PURE__ */ jsx(VisuallyHidden, { children: isTimeValue(calendarValue) ? /* @__PURE__ */ jsx(
      "input",
      {
        onChange: validationOnChange,
        onBlur,
        onInvalid,
        ...ariaProps,
        ...getNativeProps(),
        ref,
        value: calendarValue?.absoluteDate ?? "",
        tabIndex: -1
      }
    ) : /* @__PURE__ */ jsxs(Fragment, { children: [
      /* @__PURE__ */ jsx(
        "input",
        {
          onChange: validationOnChange,
          onBlur,
          onInvalid,
          ...ariaProps,
          ...getNativeProps("-from"),
          ref,
          value: isFullTimeframe(calendarValue) && calendarValue?.from?.value ? calendarValue.from.value : "",
          tabIndex: -1
        }
      ),
      /* @__PURE__ */ jsx(
        "input",
        {
          ...getNativeProps("-to"),
          ref: refTo,
          value: calendarValue?.to?.value ?? "",
          id: `${id}-to`,
          type: "hidden"
        }
      )
    ] }) })
  ] });
}
CalendarComponent["displayName"] = COMPONENT_NAME;
const CalendarRoot = forwardRef(CalendarComponent);
export {
  CalendarRoot
};
