// @dynatrace/strato-components/esm/forms/calendar/Calendar.sty.js
import "./Calendar.css";
import { createRuntimeFn as _7a468 } from "@vanilla-extract/recipes/createRuntimeFn";
var body = _7a468({ defaultClassName: "_xm8mdk3-3-6-0", variantClassNames: { isReadOnly: { true: "_xm8mdk4-3-6-0" } }, defaultVariants: {}, compoundVariants: [] });
var calendar = _7a468({ defaultClassName: "_xm8mdk0-3-6-0", variantClassNames: { readOnly: { true: "_xm8mdk1-3-6-0" } }, defaultVariants: {}, compoundVariants: [] });
var clearBackground = "_xm8mdkm-3-6-0";
var dayRow = "_xm8mdk6-3-6-0";
var disabled = "_xm8mdk2-3-6-0";
var field = _7a468({ defaultClassName: "_xm8mdk9-3-6-0", variantClassNames: { color: { primary: "_xm8mdka-3-6-0", neutral: "_xm8mdkb-3-6-0", critical: "_xm8mdkc-3-6-0" }, isIncompleteSelection: { true: "_xm8mdkd-3-6-0" }, isDisabled: { true: "_xm8mdke-3-6-0" } }, defaultVariants: { isDisabled: false, color: "neutral" }, compoundVariants: [[{ color: "primary", isDisabled: true }, "_xm8mdkf-3-6-0"], [{ color: "neutral", isDisabled: true }, "_xm8mdkg-3-6-0"], [{ color: "critical", isDisabled: true }, "_xm8mdkh-3-6-0"]] });
var monthRow = "_xm8mdk7-3-6-0";
var rangeField = _7a468({ defaultClassName: "_xm8mdki-3-6-0", variantClassNames: { rangeKind: { first: "_xm8mdkj-3-6-0", last: "_xm8mdkk-3-6-0", inbetween: "_xm8mdkl-3-6-0" } }, defaultVariants: {}, compoundVariants: [] });
var row = "_xm8mdk5-3-6-0";
var weekdays = "_xm8mdkn-3-6-0";
var yearRow = "_xm8mdk8-3-6-0";
export {
  body,
  calendar,
  clearBackground,
  dayRow,
  disabled,
  field,
  monthRow,
  rangeField,
  row,
  weekdays,
  yearRow
};
