// @dynatrace/strato-components/esm/layouts/hooks/usePaddingProps.js
import { merge } from "lodash-es";
import { spacingToCss } from "../helpers/spacingToCss.js";
const defaultPadding = {
  padding: 0,
  paddingTop: 0,
  paddingRight: 0,
  paddingBottom: 0,
  paddingLeft: 0
};
const usePaddingProps = (props, defaultPaddingProps = defaultPadding) => {
  const {
    padding,
    p,
    paddingX,
    px,
    paddingY,
    py,
    paddingTop,
    pt,
    paddingRight,
    pr,
    paddingBottom,
    pb,
    paddingLeft,
    pl
  } = props;
  const convertedDefaultPaddingProps = {
    paddingTop: spacingToCss(
      defaultPaddingProps.paddingTop ?? defaultPaddingProps.padding
    ),
    paddingLeft: spacingToCss(
      defaultPaddingProps.paddingLeft ?? defaultPaddingProps.padding
    ),
    paddingBottom: spacingToCss(
      defaultPaddingProps.paddingBottom ?? defaultPaddingProps.padding
    ),
    paddingRight: spacingToCss(
      defaultPaddingProps.paddingRight ?? defaultPaddingProps.padding
    )
  };
  return merge({}, convertedDefaultPaddingProps, {
    paddingTop: spacingToCss(
      pt ?? paddingTop ?? py ?? paddingY ?? p ?? padding
    ),
    paddingRight: spacingToCss(
      pr ?? paddingRight ?? px ?? paddingX ?? p ?? padding
    ),
    paddingBottom: spacingToCss(
      pb ?? paddingBottom ?? py ?? paddingY ?? p ?? padding
    ),
    paddingLeft: spacingToCss(
      pl ?? paddingLeft ?? px ?? paddingX ?? p ?? padding
    )
  });
};
export {
  usePaddingProps
};
