// @dynatrace/strato-components/tables/DataTable/features/Thresholds/default-comparators.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var default_comparators_exports = {};
__export(default_comparators_exports, {
  defaultComparators: () => defaultComparators
});
module.exports = __toCommonJS(default_comparators_exports);
function greaterThanComparator(ruleValue, compareValue) {
  if (typeof ruleValue === "number" && typeof compareValue === "number") {
    return compareValue > ruleValue;
  }
  return false;
}
function lessThanComparator(ruleValue, compareValue) {
  if (typeof ruleValue === "number" && typeof compareValue === "number") {
    return compareValue < ruleValue;
  }
  return false;
}
function greaterThanOrEqualToComparator(ruleValue, compareValue) {
  if (typeof ruleValue === "number" && typeof compareValue === "number") {
    return compareValue >= ruleValue;
  }
  return false;
}
function lessThanOrEqualToComparator(ruleValue, compareValue) {
  if (typeof ruleValue === "number" && typeof compareValue === "number") {
    return compareValue <= ruleValue;
  }
  return false;
}
function equalToComparator(ruleValue, compareValue) {
  if (typeof ruleValue === "number" && typeof compareValue === "number") {
    return ruleValue === compareValue;
  }
  if (typeof ruleValue === "string" && typeof compareValue === "string") {
    return ruleValue.localeCompare(compareValue) === 0;
  }
  return false;
}
function notEqualToComparator(ruleValue, compareValue) {
  if (typeof ruleValue === "number" && typeof compareValue === "number") {
    return ruleValue !== compareValue;
  }
  if (typeof ruleValue === "string" && typeof compareValue === "string") {
    return ruleValue.localeCompare(compareValue) !== 0;
  }
  return ruleValue !== compareValue;
}
const defaultComparators = {
  "greater-than": greaterThanComparator,
  "less-than": lessThanComparator,
  "greater-than-or-equal-to": greaterThanOrEqualToComparator,
  "less-than-or-equal-to": lessThanOrEqualToComparator,
  "equal-to": equalToComparator,
  "not-equal-to": notEqualToComparator
};
