// @dynatrace/strato-components/tables/DataTable/features/Thresholds/normalize-rules.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var normalize_rules_exports = {};
__export(normalize_rules_exports, {
  normalizeThresholdRules: () => normalizeThresholdRules
});
module.exports = __toCommonJS(normalize_rules_exports);
var import_makeAccessorFunction = require("../../hooks/makeAccessorFunction.js");
function makeAccessor(accessor, fallbackAccessor) {
  const appliedAccessor = accessor ?? fallbackAccessor;
  if (typeof appliedAccessor === "string") {
    if (appliedAccessor.includes(".") || appliedAccessor.includes("[")) {
      return (0, import_makeAccessorFunction.makeAccessorFunction)(appliedAccessor);
    } else {
      return (row) => {
        return row[appliedAccessor];
      };
    }
  } else if (typeof appliedAccessor === "function") {
    return appliedAccessor;
  }
  return () => void 0;
}
function normalizeThresholdRules(thresholds, fallbackAccessor) {
  const columnThresholdDefinitions = [];
  for (const threshold of thresholds) {
    const unifiedRule = {
      color: threshold.color,
      backgroundColor: "backgroundColor" in threshold ? threshold.backgroundColor : void 0,
      rules: [],
      type: "type" in threshold ? threshold.type : void 0
    };
    const normalizedRulesArray = threshold.rules !== void 0 ? threshold.rules : [
      {
        accessor: "accessor" in threshold ? threshold.accessor : void 0,
        value: "value" in threshold ? threshold.value : void 0,
        comparator: threshold.comparator
      }
    ];
    unifiedRule.rules = normalizedRulesArray.map((rule) => {
      return {
        // Accessors might be undefined in scenarios where a custom comparator function
        // is given.
        accessor: "accessor" in rule ? makeAccessor(rule.accessor, fallbackAccessor) : void 0,
        // Value might be undefined in scenarios where a custom comparator function
        // is given.
        value: "value" in rule ? rule.value : void 0,
        comparator: rule.comparator
      };
    });
    columnThresholdDefinitions.push(unifiedRule);
  }
  return columnThresholdDefinitions;
}
