// @dynatrace/strato-components/tables/DataTable/features/Thresholds/Thresholds.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Thresholds_exports = {};
__export(Thresholds_exports, {
  Thresholds: () => Thresholds,
  useThresholds: () => useThresholds
});
module.exports = __toCommonJS(Thresholds_exports);
var import_default_comparators = require("./default-comparators.js");
var import_normalize_rules = require("./normalize-rules.js");
var import_useDataTableConfig = require("../../../hooks/useDataTableConfig.js");
var import_datatable_typeguards = require("../../hooks/useTable/datatable-typeguards.js");
var import_utils = require("../../hooks/useTable/utils.js");
var import_src = require("../../tanstack-table-core-override/src/index.js");
function useThresholds(props, options) {
  const configThresholds = (0, import_useDataTableConfig.useDataTableConfigElement)("rowThresholds");
  const rowThresholds = props.rowThresholds ?? configThresholds;
  if (rowThresholds) {
    options.rowThresholds = (0, import_normalize_rules.normalizeThresholdRules)(rowThresholds);
    options.hasThresholds = true;
  } else {
    options.rowThresholds = void 0;
    options.hasThresholds = options.columns.some((column) => {
      if ((0, import_datatable_typeguards.isGroupColumnDef)(column)) {
        const groupedColumn = column;
        return groupedColumn.columns && groupedColumn.columns.some(
          (childColumn) => childColumn.thresholds && childColumn.thresholds.length > 0
        );
      }
      return column.thresholds && column.thresholds.length > 0;
    });
  }
}
function matchesThreshold(threshold, rowData) {
  return threshold.rules.every((rule) => {
    const compareValue = rule.accessor?.(rowData) ?? void 0;
    if (typeof rule.comparator === "string") {
      return import_default_comparators.defaultComparators[rule.comparator](rule.value, compareValue);
    }
    return rule.comparator(rowData);
  });
}
const Thresholds = {
  createTable(table) {
    (0, import_utils.assignRowMethods)(table, {
      getThresholdPill: (0, import_src.memo)(
        function() {
          return [table.options.rowThresholds, this.id, this.index];
        },
        function(rowThresholds) {
          if (rowThresholds === void 0 || rowThresholds.length === 0) {
            return null;
          }
          const pillColor = rowThresholds.findLast(
            (threshold) => threshold.type === "pill" && matchesThreshold(threshold, this.original)
          )?.color;
          return pillColor ?? null;
        },
        (0, import_utils.getMemoOptions)(table.options, "debugRows", "row.getThresholdPill")
      ),
      getThresholdColors: (0, import_src.memo)(
        function() {
          return [table.options.rowThresholds, this.id, this.index];
        },
        function(rowThresholds) {
          if (rowThresholds === void 0 || rowThresholds.length === 0) {
            return null;
          }
          const matchedThreshold = rowThresholds.findLast(
            (threshold) => threshold.type === "highlight" && matchesThreshold(threshold, this.original)
          );
          if (matchedThreshold) {
            return {
              color: matchedThreshold.color,
              backgroundColor: matchedThreshold.backgroundColor
            };
          }
          return null;
        },
        (0, import_utils.getMemoOptions)(table.options, "debugRows", "row.getThresholdColors")
      )
    });
  },
  createColumn(column, table) {
    column.getThresholdRules = (0, import_src.memo)(
      () => [column.columnDef.thresholds],
      (thresholds) => {
        if (thresholds === void 0 || thresholds.length === 0) {
          return null;
        }
        const columnDef = column.columnDef;
        let fallbackAccessor;
        if ((0, import_datatable_typeguards.isAccessorKeyColumnDef)(columnDef)) {
          fallbackAccessor = columnDef.accessorKey;
        }
        if ((0, import_datatable_typeguards.isAccessorFnColumnDef)(columnDef)) {
          const accessorFn = columnDef.accessorFn;
          fallbackAccessor = (row) => accessorFn(row, 0);
        }
        return (0, import_normalize_rules.normalizeThresholdRules)(thresholds, fallbackAccessor);
      },
      (0, import_utils.getMemoOptions)(table.options, "debugColumns", "column.getThresholdRules")
    );
  },
  createCell(cell, column, row) {
    cell.getThresholdColors = () => {
      const columnThresholds = column.getThresholdRules();
      const matchedColors = {
        color: void 0,
        backgroundColor: void 0
      };
      if (columnThresholds) {
        const matchedColumnThreshold = columnThresholds.findLast(
          (threshold) => matchesThreshold(threshold, row.original)
        );
        if (matchedColumnThreshold) {
          matchedColors.color = matchedColumnThreshold.color;
          matchedColors.backgroundColor = matchedColumnThreshold.backgroundColor;
          return matchedColors;
        }
      }
      const rowThresholds = row.getThresholdColors();
      if (rowThresholds) {
        matchedColors.color = rowThresholds.color;
        matchedColors.backgroundColor = rowThresholds.backgroundColor;
      }
      return matchedColors;
    };
  }
};
