// @dynatrace/strato-components/esm/charts/core/components/tooltip/ChartTooltip/ChartTooltip.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import {
  forwardRef,
  useEffect,
  useRef
} from "react";
import { createPortal } from "react-dom";
import { chartTooltip, wrapper } from "./ChartTooltip.sty.js";
import { useMergeRefs } from "../../../../../core/hooks/useMergeRefs.js";
import { useDataDtrumContext } from "../../../context/DataDtrumContext.js";
import { isTouchableDevice } from "../../../utils/is-touchable-device.js";
import {
  ActionSelectionContext
} from "../context/ActionSelectionContext.js";
import { ChartTooltipMetadataContext } from "../context/ChartTooltipMetadataContext.js";
import { useChartTooltip } from "../shared/hooks/useChartTooltip.js";
import {
  ChartTooltipSlots,
  ChartTooltipAtoms,
  ChartTooltipTemplate
} from "../types/ChartTooltip.config.js";
const defaultSelectionValue = {
  selectedItemId: "",
  actionsMenuRef: void 0,
  updateSelectedItem: () => null
};
const _ChartTooltip = forwardRef((props, forwardedRef) => {
  const {
    children,
    pinned = false,
    position,
    updatePositionWhenPinned = false,
    actionSelectionValue = defaultSelectionValue,
    onMouseLeave,
    className,
    style
  } = props;
  const tooltipRef = useRef(null);
  const mergedRefs = useMergeRefs([forwardedRef, tooltipRef]);
  const { placement, updateTooltipPosition } = useChartTooltip(
    tooltipRef.current
  );
  const { dataDtrumMask, dataDtrumAllow } = useDataDtrumContext();
  useEffect(() => {
    if (!pinned || updatePositionWhenPinned || isTouchableDevice()) {
      updateTooltipPosition(position);
    }
  }, [
    pinned,
    position.x,
    position.y,
    updatePositionWhenPinned,
    updateTooltipPosition
  ]);
  const visibility = tooltipRef.current ? "visible" : "hidden";
  const chartTooltipContent = /* @__PURE__ */ jsx(ActionSelectionContext.Provider, { value: actionSelectionValue, children: /* @__PURE__ */ jsx(ChartTooltipMetadataContext.Provider, { value: { tooltipRef }, children: /* @__PURE__ */ jsx(
    "div",
    {
      className: wrapper({ isPinned: pinned }),
      "data-testid": "chart-tooltip",
      role: "tooltip",
      ref: mergedRefs,
      "data-placement": placement,
      onMouseLeave,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      children: /* @__PURE__ */ jsx(
        "div",
        {
          className: clsx(chartTooltip, className),
          style: { ...style, visibility },
          children
        }
      )
    }
  ) }) });
  return createPortal(chartTooltipContent, document.body);
});
_ChartTooltip.displayName = "ChartTooltip";
const ChartTooltip = Object.assign(
  _ChartTooltip,
  ChartTooltipSlots,
  ChartTooltipAtoms,
  ChartTooltipTemplate
);
export {
  ChartTooltip,
  _ChartTooltip
};
