// @dynatrace/strato-components/charts/core/components/crosshair/HorizontalCrosshair.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var HorizontalCrosshair_exports = {};
__export(HorizontalCrosshair_exports, {
  HorizontalCrosshair: () => HorizontalCrosshair
});
module.exports = __toCommonJS(HorizontalCrosshair_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_shape = require("@visx/shape");
var import_lodash_es = require("lodash");
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
var import_LeftMagnifier = require("./components/LeftMagnifier.js");
var import_RightMagnifier = require("./components/RightMagnifier.js");
var import_usePoliteContainer = require("../../../../overlays/hooks/usePoliteContainer.js");
const HorizontalCrosshair = (props) => {
  const {
    y,
    leftLabel,
    rightLabel,
    width,
    height = Infinity,
    top = 0,
    leftOffset = 0,
    rightOffset = 0,
    dashed = true,
    chartBounds
  } = props;
  const lineStrokeDasharray = dashed ? "3,2" : "0,0";
  const Portal = (0, import_usePoliteContainer.usePoliteContainer)();
  const from = { x: -leftOffset, y: top + y };
  const to = { x: width - leftOffset, y: top + y };
  const strokeWidth = 1;
  const isInRange = (0, import_lodash_es.inRange)(y, 0, height + 1);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    isInRange && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_shape.Line,
      {
        role: "graphics-symbol",
        "aria-label": "crosshair",
        from,
        to,
        stroke: import_colors.default.Border.Neutral.Accent,
        strokeWidth,
        strokeDasharray: lineStrokeDasharray,
        pointerEvents: "none"
      }
    ),
    leftLabel && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(Portal, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_LeftMagnifier.LeftMagnifier,
      {
        label: leftLabel,
        mouseY: y,
        offset: leftOffset,
        chartBounds
      }
    ) }),
    rightLabel && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(Portal, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_RightMagnifier.RightMagnifier,
      {
        label: rightLabel,
        mouseY: y,
        offset: rightOffset,
        chartBounds
      }
    ) })
  ] });
};
HorizontalCrosshair["displayName"] = "HorizontalCrosshair";
