// @dynatrace/strato-components/charts/core/components/crosshair/VerticalCrosshair.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var VerticalCrosshair_exports = {};
__export(VerticalCrosshair_exports, {
  VerticalCrosshair: () => VerticalCrosshair
});
module.exports = __toCommonJS(VerticalCrosshair_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_shape = require("@visx/shape");
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
var import_BottomMagnifier = require("./components/BottomMagnifier.js");
var import_TopMagnifier = require("./components/TopMagnifier.js");
var import_usePoliteContainer = require("../../../../overlays/hooks/usePoliteContainer.js");
const VerticalCrosshair = (0, import_react.forwardRef)((props, ref) => {
  const {
    x,
    height,
    width = Infinity,
    top = 0,
    offset = 0,
    dashed = true,
    focusable,
    chartBounds,
    isBottomLabelRotated,
    isTopLabelRotated,
    topLabel,
    bottomLabel,
    onKeyDown,
    variant = "default"
  } = props;
  const Portal = (0, import_usePoliteContainer.usePoliteContainer)();
  const lineStrokeDasharray = dashed ? "3,2" : "0,0";
  const from = { x, y: top };
  const to = { x, y: top + height };
  const strokeWidth = 1;
  const isInRange = (0, import_lodash_es.inRange)(x, 0, width + 1);
  let color = import_colors.default.Background.Field.Neutral.Accent;
  if (variant === "select") {
    color = import_colors.default.Border.Primary.Accent;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    isInRange && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_shape.Line,
      {
        role: "graphics-symbol",
        "aria-label": "crosshair",
        innerRef: ref,
        from,
        to,
        stroke: color,
        strokeWidth,
        strokeDasharray: lineStrokeDasharray,
        pointerEvents: "none",
        tabIndex: focusable ? 0 : -1,
        onKeyDown
      }
    ),
    bottomLabel && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(Portal, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_BottomMagnifier.BottomMagnifier,
      {
        label: bottomLabel,
        mouseX: x,
        offset,
        color,
        chartBounds,
        isRotated: isBottomLabelRotated
      }
    ) }),
    topLabel && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(Portal, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_TopMagnifier.TopMagnifier,
      {
        label: topLabel,
        mouseX: x,
        offset,
        color,
        chartBounds,
        isRotated: isTopLabelRotated
      }
    ) })
  ] });
});
VerticalCrosshair.displayName = "VerticalCrosshair";
