// @dynatrace/strato-components/esm/charts/core/components/crosshair/HorizontalCrosshair.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { Line } from "@visx/shape";
import { inRange } from "lodash-es";
import Colors from "@dynatrace/strato-design-tokens/colors";
import { LeftMagnifier } from "./components/LeftMagnifier.js";
import { RightMagnifier } from "./components/RightMagnifier.js";
import { usePoliteContainer } from "../../../../overlays/hooks/usePoliteContainer.js";
const HorizontalCrosshair = (props) => {
  const {
    y,
    leftLabel,
    rightLabel,
    width,
    height = Infinity,
    top = 0,
    leftOffset = 0,
    rightOffset = 0,
    dashed = true,
    chartBounds
  } = props;
  const lineStrokeDasharray = dashed ? "3,2" : "0,0";
  const Portal = usePoliteContainer();
  const from = { x: -leftOffset, y: top + y };
  const to = { x: width - leftOffset, y: top + y };
  const strokeWidth = 1;
  const isInRange = inRange(y, 0, height + 1);
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    isInRange && /* @__PURE__ */ jsx(
      Line,
      {
        role: "graphics-symbol",
        "aria-label": "crosshair",
        from,
        to,
        stroke: Colors.Border.Neutral.Accent,
        strokeWidth,
        strokeDasharray: lineStrokeDasharray,
        pointerEvents: "none"
      }
    ),
    leftLabel && /* @__PURE__ */ jsx(Portal, { children: /* @__PURE__ */ jsx(
      LeftMagnifier,
      {
        label: leftLabel,
        mouseY: y,
        offset: leftOffset,
        chartBounds
      }
    ) }),
    rightLabel && /* @__PURE__ */ jsx(Portal, { children: /* @__PURE__ */ jsx(
      RightMagnifier,
      {
        label: rightLabel,
        mouseY: y,
        offset: rightOffset,
        chartBounds
      }
    ) })
  ] });
};
HorizontalCrosshair["displayName"] = "HorizontalCrosshair";
export {
  HorizontalCrosshair
};
