// @dynatrace/strato-components/esm/charts/core/components/crosshair/VerticalCrosshair.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { Line } from "@visx/shape";
import { inRange } from "lodash-es";
import { forwardRef } from "react";
import Colors from "@dynatrace/strato-design-tokens/colors";
import { BottomMagnifier } from "./components/BottomMagnifier.js";
import { TopMagnifier } from "./components/TopMagnifier.js";
import { usePoliteContainer } from "../../../../overlays/hooks/usePoliteContainer.js";
const VerticalCrosshair = forwardRef((props, ref) => {
  const {
    x,
    height,
    width = Infinity,
    top = 0,
    offset = 0,
    dashed = true,
    focusable,
    chartBounds,
    isBottomLabelRotated,
    isTopLabelRotated,
    topLabel,
    bottomLabel,
    onKeyDown,
    variant = "default"
  } = props;
  const Portal = usePoliteContainer();
  const lineStrokeDasharray = dashed ? "3,2" : "0,0";
  const from = { x, y: top };
  const to = { x, y: top + height };
  const strokeWidth = 1;
  const isInRange = inRange(x, 0, width + 1);
  let color = Colors.Background.Field.Neutral.Accent;
  if (variant === "select") {
    color = Colors.Border.Primary.Accent;
  }
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    isInRange && /* @__PURE__ */ jsx(
      Line,
      {
        role: "graphics-symbol",
        "aria-label": "crosshair",
        innerRef: ref,
        from,
        to,
        stroke: color,
        strokeWidth,
        strokeDasharray: lineStrokeDasharray,
        pointerEvents: "none",
        tabIndex: focusable ? 0 : -1,
        onKeyDown
      }
    ),
    bottomLabel && /* @__PURE__ */ jsx(Portal, { children: /* @__PURE__ */ jsx(
      BottomMagnifier,
      {
        label: bottomLabel,
        mouseX: x,
        offset,
        color,
        chartBounds,
        isRotated: isBottomLabelRotated
      }
    ) }),
    topLabel && /* @__PURE__ */ jsx(Portal, { children: /* @__PURE__ */ jsx(
      TopMagnifier,
      {
        label: topLabel,
        mouseX: x,
        offset,
        color,
        chartBounds,
        isRotated: isTopLabelRotated
      }
    ) })
  ] });
});
VerticalCrosshair.displayName = "VerticalCrosshair";
export {
  VerticalCrosshair
};
