// @dynatrace/strato-components/charts/core/components/overlay-tooltip/_to-be-removed/_OverlayTooltipWIPStory.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var OverlayTooltipWIPStory_exports = {};
__export(OverlayTooltipWIPStory_exports, {
  OverlayTooltipCustomStory: () => OverlayTooltipCustomStory,
  OverlayTooltipHeaderStory: () => OverlayTooltipHeaderStory,
  OverlayTooltipTruncationStory: () => OverlayTooltipTruncationStory,
  OverlayTooltipWIPStory: () => OverlayTooltipWIPStory,
  TooltipProgrammaticActions: () => TooltipProgrammaticActions
});
module.exports = __toCommonJS(OverlayTooltipWIPStory_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_DummyInteractionsOverlay = require("./DummyInteractionsOverlay.js");
var import_HeaderTestTooltip = require("./HeaderTestTooltip.js");
var import_MyTooltip = require("./MyTooltip.js");
var import_Button = require("../../../../../buttons/button/Button.js");
var import_Select = require("../../../../../forms/select/Select.js");
var import_Flex = require("../../../../../layouts/flex/Flex.js");
var import_ToastContainer = require("../../../../../notifications/toast/ToastContainer.js");
var import_OverlayContext = require("../contexts/OverlayContext.js");
var import_TooltipContext = require("../contexts/TooltipContext.js");
var import_useTooltipReducer = require("../useTooltipReducer.js");
var import_TruncationStory = require("./_TruncationStory.js");
const TooltipProgrammaticActions = () => {
  const dispatch = (0, import_useTooltipReducer.useTooltipReducer)();
  const tooltip = (0, import_TooltipContext.useTooltip)();
  let snapshot = tooltip.getState();
  const reset = () => {
    snapshot = tooltip.getState();
    dispatch({ type: "RESET_TOOLTIP" })();
  };
  const restore = () => {
    snapshot.pinned && tooltip.setState(snapshot);
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { gap: 4, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button, { variant: "emphasized", onClick: reset, children: "Reset Tooltip" }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button, { variant: "emphasized", onClick: restore, children: "Restore tooltip" })
  ] });
};
const createPinAllHandler = (sub) => () => {
  sub.forEach((action) => {
    action();
  });
};
const usePinAllSetup = () => {
  const sub = /* @__PURE__ */ new Set();
  const onClick = createPinAllHandler(sub);
  return { sub, onClick };
};
const chunkArray = (array, size) => {
  const rows = [];
  for (let i = 0; i < array.length; i += size) {
    rows.push(array.slice(i, i + size));
  }
  return rows;
};
const TooltipGrid = (props) => {
  const {
    sub,
    items,
    itemsPerRow = 3,
    width = "350px",
    minHeight = "200px",
    symbolAlignment
  } = props;
  const rows = chunkArray(items, itemsPerRow);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, { children: rows.map((row, rowIndex) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Flex.Flex, { flexDirection: "row", children: row.map((item, itemIndex) => {
    const key = rowIndex * itemsPerRow + itemIndex;
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { flexDirection: "column", "data-testid": item.testId, children: [
      item.description,
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        Content,
        {
          sub,
          width,
          minHeight,
          getData: item.data,
          children: symbolAlignment && (0, import_react.isValidElement)(item.template) ? (0, import_react.cloneElement)(
            item.template,
            { symbolAlignment }
          ) : item.template
        }
      )
    ] }, key);
  }) }, rowIndex)) });
};
const TogglePinned = (props) => {
  const dispatch = (0, import_useTooltipReducer.useTooltipReducer)();
  const isPinned = (0, import_react.useRef)(false);
  (0, import_react.useEffect)(() => {
    const cb = () => {
      if (isPinned.current) {
        dispatch({
          type: "RESET_TOOLTIP"
        })();
      } else {
        dispatch({
          type: "SET_PINNED",
          payload: {
            pinned: true,
            absoluteX: 10,
            absoluteY: 30,
            data: props.getData ? props.getData() : import_MyTooltip.TooltipData[0].data()
          }
        })();
      }
      isPinned.current = !isPinned.current;
    };
    props.sub.add(cb);
    return () => {
      props.sub.delete(cb);
    };
  }, []);
  return null;
};
const Content = (props) => {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayContext.OverlayProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_TooltipContext.TooltipProvider, { children: [
    props.actions && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(TooltipProgrammaticActions, {}),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(TogglePinned, { sub: props.sub, getData: props.getData }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "div",
      {
        style: {
          width: props.width || "200px",
          height: "100%",
          maxWidth: "100%",
          maxHeight: "100%",
          resize: "both",
          minHeight: props.minHeight || "50px",
          borderRight: "1px dashed #DDA0DD",
          overflow: "hidden",
          position: "relative"
        },
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_DummyInteractionsOverlay.DummyInteractionsOverlay,
          {
            getData: props.getData || import_MyTooltip.TooltipData[0].data,
            children: props.children
          }
        )
      }
    )
  ] }) });
};
const Selector = function(props) {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Select.Select,
    {
      value: props.value,
      onChange: (newValue) => {
        props.update(newValue);
      },
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Select.Select.Content, { children: props.options.map((option) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Select.Select.Option, { value: option, children: `${option}` }, String(option))) })
    }
  );
};
const OverlayTooltipCustomStory = () => {
  const [nTooltips, setNTooltips] = (0, import_react.useState)(5);
  const { sub, onClick } = usePinAllSetup();
  const Shared = () => {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        Selector,
        {
          options: [5, 50, 100, 150, 300],
          value: nTooltips,
          update: setNTooltips
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button, { onClick, children: "Pin all" })
    ] });
  };
  if (nTooltips === 5) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      import_Flex.Flex,
      {
        flexWrap: "nowrap",
        gap: 6,
        style: { marginTop: "10px" },
        flexDirection: "column",
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(Shared, {}),
          Array.from({ length: 5 }).map((_, index) => {
            return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              Content,
              {
                sub,
                actions: true,
                width: "100%",
                minHeight: "400px",
                children: import_MyTooltip.TooltipData[0].template
              },
              index
            );
          })
        ]
      }
    );
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(Shared, {}),
    Array.from({ length: nTooltips / 10 }).map((_, index) => {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_Flex.Flex,
        {
          gap: 6,
          style: { marginTop: "10px" },
          flexDirection: "row",
          children: Array.from({ length: 10 }).map((_2, index2) => {
            return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(Content, { sub, children: import_MyTooltip.TooltipData[0].template }, index2);
          })
        },
        index
      );
    })
  ] });
};
const OverlayTooltipWIPStory = (props) => {
  const { sub, onClick } = usePinAllSetup();
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ToastContainer.ToastContainer, {}),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button, { onClick, children: "Pin all" }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      TooltipGrid,
      {
        sub,
        items: import_MyTooltip.TooltipData,
        symbolAlignment: props?.symbolAlignment
      }
    )
  ] });
};
const OverlayTooltipHeaderStory = () => {
  const { sub, onClick } = usePinAllSetup();
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button, { onClick, children: "Pin all" }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(TooltipGrid, { sub, items: import_HeaderTestTooltip.HeaderTestData })
  ] });
};
const OverlayTooltipTruncationStory = () => {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TruncationStory.TruncationStory, {});
};
