// @dynatrace/strato-components/charts/core/components/overlay-tooltip/_to-be-removed/actions.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var actions_exports = {};
__export(actions_exports, {
  asyncSeriesActions: () => asyncSeriesActions
});
module.exports = __toCommonJS(actions_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_Toast = require("../../../../../notifications/toast/Toast.js");
var import_ChartSeriesAction = require("../../series-action/ChartSeriesAction.js");
const TOASTER_LIFESPAN = 800;
const IS_ASYNC = true;
let toasterCounter = 0;
const seriesActions = (data) => {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_ChartSeriesAction.ChartSeriesAction, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      import_ChartSeriesAction.ChartSeriesAction.Item,
      {
        onSelect: () => {
          (0, import_Toast.showToast)({
            title: `Inspecting ${data.name} ...`,
            type: "info",
            lifespan: 1e3
          });
        },
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ChartSeriesAction.ChartSeriesAction.ItemIcon, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.MagnifyingGlassIcon, {}) }),
          "Inspect"
        ]
      },
      "custom-action"
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      import_ChartSeriesAction.ChartSeriesAction.Item,
      {
        disabled: true,
        onSelect: () => void 0,
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ChartSeriesAction.ChartSeriesAction.ItemIcon, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.DatabaseIcon, {}) }),
          "A disabled action"
        ]
      },
      "disabled-action"
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_ChartSeriesAction.ChartSeriesAction.Intent,
      {
        options: {
          responseProperties: ["dt.share"]
        },
        payload: { "dt.query": "describe logs" },
        onResponse: (response) => {
          if (response.status === "success") {
            (0, import_Toast.showToast)({
              title: "done...",
              type: "info",
              lifespan: TOASTER_LIFESPAN
            });
          } else {
            (0, import_Toast.showToast)({
              title: response.error.message,
              type: "critical",
              lifespan: TOASTER_LIFESPAN
            });
          }
        },
        children: "Send intent with response"
      }
    )
  ] });
};
const asyncSeriesActions = (data) => {
  if (!IS_ASYNC) {
    return seriesActions(data);
  }
  return new Promise((resolve, reject) => {
    (0, import_Toast.showToast)({
      title: `Loading series actions ...: ${toasterCounter++}`,
      type: "warning",
      lifespan: TOASTER_LIFESPAN
    });
    setTimeout(() => {
      if (toasterCounter % 5 === 0) {
        (0, import_Toast.showToast)({
          title: `Failed to load series actions ...: ${toasterCounter++}`,
          type: "critical",
          lifespan: TOASTER_LIFESPAN
        });
        reject(new Error("Failed to load actions"));
      } else {
        resolve(seriesActions(data));
      }
    }, 500);
  });
};
