// @dynatrace/strato-components/charts/core/components/overlay-tooltip/_to-be-removed/DummyInteractionsOverlay.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DummyInteractionsOverlay_exports = {};
__export(DummyInteractionsOverlay_exports, {
  DummyInteractionsOverlay: () => DummyInteractionsOverlay
});
module.exports = __toCommonJS(DummyInteractionsOverlay_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_lodash_es = require("lodash");
var import_constants = require("../../../constants.js");
var import_ChartGrid_css = require("../../chart-grid/ChartGrid.sty.js");
var import_HorizontalCrosshair = require("../../crosshair/HorizontalCrosshair.js");
var import_VerticalCrosshair = require("../../crosshair/VerticalCrosshair.js");
var import_InteractionsOverlay_css = require("../../interactions-overlay/InteractionsOverlay.sty.js");
var import_OverlayContext = require("../contexts/OverlayContext.js");
var import_TooltipContext = require("../contexts/TooltipContext.js");
var import_selectors = require("../selectors.js");
var import_useTooltipReducer = require("../useTooltipReducer.js");
var import_utils = require("../utils.js");
const DummyInteractionsOverlay = (props) => {
  const dispatch = (0, import_useTooltipReducer.useTooltipReducer)();
  const tooltip = (0, import_TooltipContext.useTooltip)();
  const overlay = (0, import_OverlayContext.useOverlay)();
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    "div",
    {
      role: "none",
      ref: overlay.overlayRef,
      className: import_InteractionsOverlay_css.interactionsOverlay,
      onPointerMove: (event) => (0, import_utils.isOwnEvent)(event) && dispatch({
        type: "SET_POSITION",
        payload: {
          data: props.getData(),
          absoluteX: Math.round(event.clientX),
          absoluteY: Math.round(event.clientY)
        }
      })(),
      onPointerLeave: dispatch({ type: "SET_OUT_BOUNDS" }),
      onClick: (event) => {
        if (!(0, import_utils.isOwnEvent)(event)) {
          return;
        }
        if (tooltip.getState().pinned) {
          dispatch({
            type: "SET_PINNED",
            payload: {
              pinned: false,
              data: {
                geometry: "none",
                series: void 0,
                header1: void 0,
                subheader: void 0
              }
            }
          })();
        } else {
          dispatch({
            type: "SET_PINNED",
            payload: {
              pinned: true,
              data: props.getData(),
              absoluteX: Math.round(event.clientX),
              absoluteY: Math.round(event.clientY)
            }
          })();
        }
      },
      style: {
        background: "#E6E6FA",
        position: "absolute",
        width: "98%",
        height: "98%",
        top: "1%",
        left: "1%"
      },
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(Crosshair, {}),
        props.children
      ]
    }
  );
};
const Crosshair = () => {
  const overlay = (0, import_OverlayContext.useOverlay)();
  (0, import_selectors.useSelector)(
    ({ absoluteX, absoluteY, pinned: pinned2 }) => `${absoluteX}-${absoluteY}-${pinned2}`
  );
  const { relativeX, relativeY, pinned } = (0, import_selectors.useTooltipSnapshot)();
  const bounds = overlay.getBounds();
  if ((0, import_lodash_es.isNil)(bounds && relativeX && relativeY)) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    "svg",
    {
      className: (0, import_clsx.default)(
        import_ChartGrid_css.plotLayout,
        {
          overflow: "visible",
          cursor: "crosshair",
          zIndex: import_constants.ChartOverlaysElevation.crosshair,
          pointerEvents: "none"
        },
        import_InteractionsOverlay_css.interactionsOverlay
      ),
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_HorizontalCrosshair.HorizontalCrosshair,
          {
            dashed: !pinned,
            y: relativeY,
            width: bounds?.width,
            height: bounds?.height,
            chartBounds: bounds,
            leftOffset: 0,
            rightOffset: 0
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_VerticalCrosshair.VerticalCrosshair,
          {
            dashed: !pinned,
            x: relativeX,
            height: bounds?.height,
            width: bounds?.width,
            chartBounds: bounds
          }
        )
      ]
    }
  );
};
