// @dynatrace/strato-components/charts/core/components/overlay-tooltip/_to-be-removed/MyTooltip.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MyTooltip_exports = {};
__export(MyTooltip_exports, {
  DotTooltipData: () => DotTooltipData,
  TooltipData: () => TooltipData,
  XYDualAxisTooltip: () => XYDualAxisTooltip
});
module.exports = __toCommonJS(MyTooltip_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_actions = require("./actions.js");
var import_utils = require("./utils.js");
var import_Toast = require("../../../../../notifications/toast/Toast.js");
var import_OverlayTooltip = require("../OverlayTooltip.js");
var import_utils2 = require("../utils.js");
const XYTooltip = (props) => {
  const data = (0, import_utils.useTooltipData)().data;
  if ((0, import_utils2.isTooltipDataEmpty)(data)) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_OverlayTooltip.OverlayTooltip, { symbolAlignment: props.symbolAlignment, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayTooltip.OverlayTooltip.Header, { header1: data.header1, label: props.label }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_OverlayTooltip.OverlayTooltip.Series,
      {
        activeItem: 0,
        variant: props.variant,
        data: data.series,
        seriesItemTemplate: (current) => {
          return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_OverlayTooltip.OverlayTooltip.SeriesItem,
            {
              ...current,
              relativeValueUnit: "%",
              actions: () => (0, import_actions.asyncSeriesActions)(current)
            }
          );
        }
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_OverlayTooltip.OverlayTooltip.Action,
      {
        label: "Copy all values",
        icon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.CopyIcon, {}),
        onSelect: () => (0, import_Toast.showToast)({
          title: "Copy all values",
          type: "info",
          lifespan: 1e3
        })
      }
    )
  ] });
};
const PieTooltip = (props) => {
  const data = (0, import_utils.useTooltipData)().data;
  if ((0, import_utils2.isTooltipDataEmpty)(data)) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayTooltip.OverlayTooltip, { symbolAlignment: props.symbolAlignment, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_OverlayTooltip.OverlayTooltip.SeriesItem,
    {
      ...data.series[0],
      relativeValueUnit: "%",
      actions: () => (0, import_actions.asyncSeriesActions)(data.series[0])
    }
  ) });
};
const MeterBarTooltip = (props) => {
  const data = (0, import_utils.useTooltipData)().data;
  if ((0, import_utils2.isTooltipDataEmpty)(data)) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_OverlayTooltip.OverlayTooltip, { symbolAlignment: props.symbolAlignment, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_OverlayTooltip.OverlayTooltip.SeriesHeader,
      {
        seriesLabel: "App rating",
        valueLabel: "Amount"
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_OverlayTooltip.OverlayTooltip.SeriesItem,
      {
        ...data.series[0],
        actions: () => (0, import_actions.asyncSeriesActions)(data.series[0])
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_OverlayTooltip.OverlayTooltip.Action,
      {
        label: "Only label action 1",
        onSelect: () => (0, import_Toast.showToast)({
          title: "Only label action 1",
          type: "info",
          lifespan: 1e3
        })
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_OverlayTooltip.OverlayTooltip.Action,
      {
        label: "Only label action 2",
        onSelect: () => (0, import_Toast.showToast)({
          title: "Only label action 2",
          type: "info",
          lifespan: 1e3
        })
      }
    )
  ] });
};
const TreeMapTooltip = (props) => {
  const data = (0, import_utils.useTooltipData)().data;
  if ((0, import_utils2.isTooltipDataEmpty)(data)) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_OverlayTooltip.OverlayTooltip, { symbolAlignment: props.symbolAlignment, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayTooltip.OverlayTooltip.Header, { header1: data.header1 }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_OverlayTooltip.OverlayTooltip.SeriesHeader,
      {
        seriesLabel: "Country",
        valueLabel: "Inhabitants"
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayTooltip.OverlayTooltip.SeriesItem, { ...data.series[0] })
  ] });
};
const TreeMapParentTooltip = (props) => {
  const data = (0, import_utils.useTooltipData)().data;
  if ((0, import_utils2.isTooltipDataEmpty)(data)) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_OverlayTooltip.OverlayTooltip, { symbolAlignment: props.symbolAlignment, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayTooltip.OverlayTooltip.Header, { header1: data.header1 }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_OverlayTooltip.OverlayTooltip.SeriesHeader,
      {
        seriesLabel: "Region",
        valueLabel: "Inhabitants"
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_OverlayTooltip.OverlayTooltip.SeriesItem,
      {
        ...data.series[0],
        actions: () => (0, import_actions.asyncSeriesActions)(data.series[0])
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayTooltip.OverlayTooltip.ActionsPlaceholder, { text: "Click to access actions" }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_OverlayTooltip.OverlayTooltip.Action,
      {
        label: "Copy all values",
        icon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.CopyIcon, {}),
        onSelect: () => (0, import_Toast.showToast)({
          title: "Copy all values",
          type: "info",
          lifespan: 1e3
        })
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_OverlayTooltip.OverlayTooltip.Action,
      {
        label: "Download data",
        icon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.DownloadIcon, {}),
        onSelect: () => (0, import_Toast.showToast)({
          title: "Download data",
          type: "info",
          lifespan: 1e3
        })
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_OverlayTooltip.OverlayTooltip.Action,
      {
        label: "View details",
        onSelect: () => (0, import_Toast.showToast)({
          title: "View details",
          type: "info",
          lifespan: 1e3
        })
      }
    )
  ] });
};
const HoneyCombTooltip = (props) => {
  const data = (0, import_utils.useTooltipData)().data;
  if ((0, import_utils2.isTooltipDataEmpty)(data)) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_OverlayTooltip.OverlayTooltip, { symbolAlignment: props.symbolAlignment, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_OverlayTooltip.OverlayTooltip.SeriesHeader,
      {
        seriesLabel: "Severity",
        valueLabel: "Incidents"
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_OverlayTooltip.OverlayTooltip.SeriesItem,
      {
        ...data.series[0],
        actions: () => (0, import_actions.asyncSeriesActions)(data.series[0])
      }
    )
  ] });
};
const CustomIconTooltip = (props) => {
  const data = (0, import_utils.useTooltipData)().data;
  if ((0, import_utils2.isTooltipDataEmpty)(data)) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_OverlayTooltip.OverlayTooltip, { symbolAlignment: props.symbolAlignment, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayTooltip.OverlayTooltip.SeriesHeader, { seriesLabel: "Category", valueLabel: "Count" }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_OverlayTooltip.OverlayTooltip.Series,
      {
        activeItem: 0,
        data: data.series,
        seriesItemTemplate: (current) => {
          return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_OverlayTooltip.OverlayTooltip.SeriesItem,
              {
                icon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.CopyIcon, {}),
                name: "Copy actions",
                value: "Copy all values"
              }
            ),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_OverlayTooltip.OverlayTooltip.SeriesItem,
              {
                icon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.MagnifyingGlassIcon, {}),
                name: "View details",
                value: "View"
              }
            ),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_OverlayTooltip.OverlayTooltip.SeriesItem,
              {
                icon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.DownloadIcon, {}),
                name: "Download",
                value: "Download data"
              }
            )
          ] });
        }
      }
    )
  ] });
};
const CustomSymbolEmojiTooltip = (props) => {
  const data = (0, import_utils.useTooltipData)().data;
  if ((0, import_utils2.isTooltipDataEmpty)(data)) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_OverlayTooltip.OverlayTooltip, { symbolAlignment: props.symbolAlignment, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayTooltip.OverlayTooltip.SeriesHeader, { seriesLabel: "Category", valueLabel: "Count" }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_OverlayTooltip.OverlayTooltip.Series,
      {
        activeItem: 0,
        data: data.series,
        seriesItemTemplate: (current) => {
          return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_OverlayTooltip.OverlayTooltip.SeriesItem,
              {
                symbol: "\u{1F680}",
                name: "Launches",
                value: "42"
              }
            ),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayTooltip.OverlayTooltip.SeriesItem, { symbol: "\u{1F30D}", name: "Regions", value: "7" }),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayTooltip.OverlayTooltip.SeriesItem, { symbol: "A", name: "Alpha", value: "100" }),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayTooltip.OverlayTooltip.SeriesItem, { symbol: "\u2605", name: "Star", value: "200" }),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayTooltip.OverlayTooltip.SeriesItem, { symbol: "#", name: "Hash", value: "300" })
          ] });
        }
      }
    )
  ] });
};
const CustomShapeIconsTooltip = (props) => {
  const data = (0, import_utils.useTooltipData)().data;
  if ((0, import_utils2.isTooltipDataEmpty)(data)) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_OverlayTooltip.OverlayTooltip, { symbolAlignment: props.symbolAlignment, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayTooltip.OverlayTooltip.SeriesHeader, { seriesLabel: "Shape", valueLabel: "Value" }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_OverlayTooltip.OverlayTooltip.Series,
      {
        activeItem: 0,
        data: data.series,
        seriesItemTemplate: (current) => {
          return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_OverlayTooltip.OverlayTooltip.SeriesItem,
              {
                color: "#3DD598",
                shape: "circle",
                name: "Circle (default)",
                value: "100"
              }
            ),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_OverlayTooltip.OverlayTooltip.SeriesItem,
              {
                color: "#EB6B52",
                shape: "square",
                name: "Square",
                value: "200"
              }
            ),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_OverlayTooltip.OverlayTooltip.SeriesItem,
              {
                color: "#63C45E",
                shape: "triangle",
                name: "Triangle",
                value: "300"
              }
            ),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_OverlayTooltip.OverlayTooltip.SeriesItem,
              {
                color: "#EB6B52",
                shape: "bubble",
                name: "Bubble",
                value: "700"
              }
            )
          ] });
        }
      }
    )
  ] });
};
const XYDualAxisTooltip = (props) => {
  const data = (0, import_utils.useTooltipData)().data;
  if ((0, import_utils2.isTooltipDataEmpty)(data)) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_OverlayTooltip.OverlayTooltip, { symbolAlignment: props.symbolAlignment, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_OverlayTooltip.OverlayTooltip.Header,
      {
        header1: data.header1,
        header2: [/* @__PURE__ */ new Date("2026-01-19 10:30"), /* @__PURE__ */ new Date("2025-02-10 10:55")],
        subheader: data.subheader,
        showRangeDuration1: true
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_OverlayTooltip.OverlayTooltip.Series,
      {
        activeItem: 0,
        variant: "single",
        data: data.series,
        seriesItemTemplate: (current) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayTooltip.OverlayTooltip.SeriesItem, { ...current })
      }
    )
  ] });
};
const DotTooltipData = {
  geometry: "line",
  header1: ["Q3"],
  series: [
    {
      name: "R&D",
      color: "#3DD598",
      value: "505K",
      valueUnit: "$"
    },
    {
      name: "Support",
      color: "#EB6B52",
      value: "350K",
      valueUnit: "$",
      relativeValue: void 0,
      relativeValueUnit: void 0
    },
    {
      name: "Sales",
      color: "#63C45E",
      value: "320K",
      valueUnit: "$",
      relativeValue: void 0,
      relativeValueUnit: void 0
    },
    {
      name: "Marketing",
      color: "#D971C0",
      value: "305K",
      valueUnit: "$",
      relativeValue: void 0,
      relativeValueUnit: void 0
    },
    {
      name: "HR",
      color: "#D9B671",
      value: "205K",
      valueUnit: "$",
      relativeValue: void 0,
      relativeValueUnit: void 0
    },
    {
      name: "IT",
      color: "#3DD598",
      value: "105K",
      valueUnit: "$"
    },
    {
      name: "Finance",
      color: "#EB6B52",
      value: "95K",
      valueUnit: "$"
    },
    {
      name: "Admin",
      color: "#63C45E",
      value: "80K",
      valueUnit: "$"
    },
    {
      name: "Management",
      color: "#D971C0",
      value: "75K",
      valueUnit: "$"
    },
    {
      name: "Legal",
      color: "#D9B671",
      value: "65K",
      valueUnit: "$"
    },
    {
      name: "Operations",
      color: "#3DD598",
      value: "55K",
      valueUnit: "$"
    },
    {
      name: "Other",
      color: "#EB6B52",
      value: "50K",
      valueUnit: "$"
    },
    {
      name: "Logistics",
      color: "#63C45E",
      value: "45K",
      valueUnit: "$"
    },
    {
      name: "Procurement",
      color: "#D971C0",
      value: "30K",
      valueUnit: "$"
    }
  ]
};
const rectTooltipData = {
  geometry: "rect",
  header1: [new Date(2023, 3, 1, 7, 30), new Date(2023, 7, 1, 7, 35)],
  series: [
    {
      name: "R&D",
      color: "#3DD598",
      value: "505K $",
      relativeValue: "25"
    }
  ]
};
const rectTooltipNumericData = {
  geometry: "rect",
  header1: [1325, 1388],
  series: [
    {
      name: "HR",
      color: "#D9B671",
      value: "205K $",
      relativeValue: void 0
    }
  ]
};
const headerStyle = { margin: 0, marginBlock: "unset" };
const TooltipData = [
  {
    testId: "dot-with-single-series",
    description: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("h3", { style: headerStyle, children: "Dot tooltip with single series" }),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { children: "1 Global Action " })
    ] }),
    data: () => DotTooltipData,
    template: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(XYTooltip, { label: "Quarter", variant: "single" })
  },
  {
    testId: "dot-with-timeframe",
    description: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("h3", { style: headerStyle, children: "Dot tooltip with timeframe" }),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { children: "1 Global Action " })
    ] }),
    data: () => ({
      ...DotTooltipData,
      header1: [new Date(2023, 0, 1, 7, 30), new Date(2023, 0, 1, 7, 35)]
    }),
    template: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(XYTooltip, { variant: "single", label: "" })
  },
  {
    testId: "rect-with-single-series",
    description: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("h3", { style: headerStyle, children: "Rect tooltip with single series" }),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { children: "1 Global Action " })
    ] }),
    data: () => ({
      geometry: "rect",
      header1: [new Date(2023, 0, 1, 7, 30), new Date(2023, 0, 1, 7, 35)],
      series: [
        {
          name: "WebServiceA-CW1-1",
          color: "#3DD598",
          value: 35,
          valueUnit: "ms"
        }
      ]
    }),
    template: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(XYTooltip, { variant: "single" })
  },
  {
    testId: "dot-multiple-series",
    description: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("h3", { style: headerStyle, children: "Dot tooltip with multiple series" }),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { children: "1 Global Action " })
    ] }),
    data: () => DotTooltipData,
    template: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(XYTooltip, { variant: "shared" })
  },
  {
    testId: "rect-with-timeframe",
    description: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("h3", { style: headerStyle, children: "Rect tooltip with timeframe" }),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { children: "1 Global Action " })
    ] }),
    data: () => rectTooltipData,
    template: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(XYTooltip, { variant: "single" })
  },
  {
    testId: "rect-with-numeric-range",
    description: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("h3", { style: headerStyle, children: "Rect tooltip with numeric range" }),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { children: "1 Global Action " })
    ] }),
    data: () => rectTooltipNumericData,
    template: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(XYTooltip, { variant: "single", label: "" })
  },
  {
    testId: "pie",
    description: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("h3", { children: "Pie tooltip" }),
    data: () => ({
      geometry: "pie",
      series: [
        {
          name: "Latam",
          color: "#3DD598",
          value: "35K",
          valueUnit: "$",
          relativeValue: "25"
        }
      ]
    }),
    template: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(PieTooltip, {})
  },
  {
    testId: "meter-bar",
    description: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("h3", { style: headerStyle, children: "MeterBar tooltip" }),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { children: "2 Global Actions (only label)" })
    ] }),
    data: () => ({
      geometry: "meterbar",
      series: [
        {
          name: "Good",
          color: "#3DD598",
          value: "85"
        }
      ]
    }),
    template: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(MeterBarTooltip, {})
  },
  {
    testId: "treeMap-with-parent",
    description: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("h3", { style: headerStyle, children: "TreeMap tooltip with parent" }),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { children: "3 Global Actions" })
    ] }),
    data: () => ({
      geometry: "treemap",
      header1: ["United States"],
      series: [
        {
          name: "New York City",
          color: "#3DD598",
          value: "18.8M"
        }
      ]
    }),
    template: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(TreeMapParentTooltip, {})
  },
  {
    testId: "treeMap-without-parent",
    description: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("h3", { children: "TreeMap tooltip without parent" }),
    data: () => ({
      geometry: "treemap",
      header1: [],
      series: [
        {
          name: "Brazil",
          color: "#3DD598",
          value: "35.2M"
        }
      ]
    }),
    template: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(TreeMapTooltip, {})
  },
  {
    testId: "treeMap-without-parent-no-color",
    description: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("h3", { children: "TreeMap tooltip without parent and no color" }),
    data: () => ({
      geometry: "treemap",
      header1: [],
      series: [
        {
          name: "Brazil",
          value: "35.2M"
        }
      ]
    }),
    template: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(TreeMapTooltip, {})
  },
  {
    testId: "honeyComb",
    description: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("h3", { children: "HoneyComb tooltip" }),
    data: () => ({
      geometry: "honeycomb",
      header1: [],
      series: [
        {
          name: "Emergency",
          color: "#3DD598",
          value: "4.1K"
        }
      ]
    }),
    template: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(HoneyCombTooltip, {})
  },
  {
    description: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("h3", { children: "Custom symbol with Icons" }),
    testId: "custom-symbol-with-icons",
    data: () => ({
      geometry: "dot",
      header1: [],
      series: [
        {
          name: "Launches",
          color: "",
          value: "42"
        }
      ]
    }),
    template: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(CustomIconTooltip, {})
  },
  {
    testId: "custom-symbol-with-emojis-and-characters",
    description: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("h3", { children: "Custom symbol emojis & characters" }),
    data: () => ({
      geometry: "dot",
      header1: [],
      series: [
        {
          name: "Launches",
          color: "",
          value: "42"
        }
      ]
    }),
    template: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(CustomSymbolEmojiTooltip, {})
  },
  {
    testId: "shape-icons-circle-square-triangle-etc",
    description: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("h3", { children: "Shape icons (circle, square, triangle, etc.)" }),
    data: () => ({
      geometry: "dot",
      header1: [],
      series: [
        {
          name: "Circle",
          color: "#3DD598",
          value: "100"
        }
      ]
    }),
    template: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(CustomShapeIconsTooltip, {})
  },
  {
    testId: "xy-dual-axis-header",
    description: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("h3", { style: headerStyle, children: "XY tooltip with dual axis header" }),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { children: "Both X and Y axis ranges with duration" })
    ] }),
    data: () => ({
      geometry: "rect",
      header1: [/* @__PURE__ */ new Date("2023-01-19"), /* @__PURE__ */ new Date("2023-02-10")],
      subheader: ["18 data points across 2 series"],
      series: [
        {
          name: "WebServiceA",
          color: "#3DD598",
          value: "65.0",
          valueUnit: "ms"
        }
      ]
    }),
    template: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(XYDualAxisTooltip, {})
  }
];
