// @dynatrace/strato-components/esm/forms/spin-buttons/text-spin-button/TextSpinButton.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import {
  forwardRef,
  useEffect,
  useImperativeHandle,
  useRef,
  useState
} from "react";
import { isElement } from "react-is";
import { useValueNavigation } from "./hooks/useValueNavigation.js";
import { useAriaLabelingProps } from "../../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../../core/hooks/useBehavioralTrackingProps.js";
import { useSelection } from "../hooks/useSelection.js";
import { useSpinButtonInteraction } from "../hooks/useSpinButtonInteraction.js";
import { spinButtonCSS } from "../SpinButton.sty.js";
function isTextSpinButton(node) {
  return node !== void 0 && node !== null && isElement(node) && node.type === TextSpinButton;
}
const TextSpinButton = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const {
    placeholder,
    disabled = false,
    readOnly = false,
    options,
    required = false,
    hasError,
    className: consumerClassName,
    style: consumerStyle,
    onKeyDownHandler: propOnKeyDownHandler,
    "data-testid": dataTestId,
    onChange,
    "data-pattern": dataPattern,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps] = useBehavioralTrackingProps(remainingProps);
  const ariaProps = useAriaLabelingProps(props);
  const rootRef = useRef(null);
  const { getSelection, setSelection, selectAll, deselectAll } = useSelection(rootRef);
  const [error, setError] = useState(false);
  const { value, onKeyDownHandler, searchValue, clearValue } = useValueNavigation(props);
  useImperativeHandle(
    forwardedRef,
    () => {
      return {
        type: "text",
        element: rootRef,
        setSelection,
        getSelection,
        selectAll,
        deselectAll,
        searchValue,
        clearValue,
        onChange
      };
    },
    [
      clearValue,
      deselectAll,
      getSelection,
      onChange,
      searchValue,
      selectAll,
      setSelection
    ]
  );
  useEffect(() => {
    if (!rootRef.current) {
      return;
    }
    rootRef.current.textContent = value ?? "";
    document.activeElement === rootRef.current && selectAll();
  }, [value]);
  useEffect(() => {
    if (required && value === null) {
      setError(true);
    } else {
      setError(value !== null && !options.includes(value));
    }
  }, [required, value, options]);
  const { onClick, onFocus, onBlur } = useSpinButtonInteraction({
    disabled,
    selectAll,
    deselectAll,
    rootElement: rootRef.current
  });
  return /* @__PURE__ */ jsx(
    "span",
    {
      "data-testid": dataTestId,
      contentEditable: (
        // contentEditable when focused and non-editable when blurred, allowing text selection.
        !disabled && !readOnly ? true : void 0
      ),
      ref: rootRef,
      tabIndex: disabled ? void 0 : 0,
      role: "spinbutton",
      spellCheck: false,
      className: clsx(
        consumerClassName,
        spinButtonCSS({ disabled, error: hasError ?? error })
      ),
      style: {
        "--data-placeholder": `"${placeholder}"`,
        ...consumerStyle
      },
      "aria-disabled": disabled,
      ...ariaProps,
      "aria-label": ariaProps["aria-label"] ?? placeholder,
      onKeyDown: propOnKeyDownHandler ?? onKeyDownHandler,
      onClick,
      onFocus,
      onBlur,
      onBeforeInput: (e) => e.preventDefault(),
      onPaste: (e) => e.preventDefault(),
      onDrop: (e) => e.preventDefault(),
      onCut: (e) => e.preventDefault(),
      "data-pattern": dataPattern,
      ...behavioralTrackingProps
    }
  );
});
TextSpinButton.displayName = "_TextSpinButton";
export {
  TextSpinButton,
  isTextSpinButton
};
