// @dynatrace/strato-components/esm/charts/core/components/axes/Axis.js
import { jsx } from "react/jsx-runtime";
import { Axis as VisxAxis } from "@visx/axis";
import { axis } from "./Axis.sty.js";
import {
  AXIS_TICK_LENGTH,
  axisFontBaseProps,
  axisFontProps,
  DEFAULT_AXIS_FONT_COLOR
} from "./constants.js";
import {
  charsFontWidth14pixelsSize400Map,
  charsFontWidth12pixelsSize400Map
} from "../text-ellipsis/chars-font-width-map.js";
import { truncateString } from "../text-ellipsis/truncate-string.js";
const labelProps = { ...axisFontProps, textAnchor: "middle" };
const labelBaseProps = {
  ...axisFontBaseProps,
  textAnchor: "middle"
};
const Axis = ({
  width = "100%",
  height = "100%",
  containerSize,
  label,
  size = "small",
  tickLabelProps = (...args) => ({}),
  tickColor,
  ...restProps
}) => {
  const fontProps = {
    ...size === "small" ? axisFontProps : axisFontBaseProps,
    fill: tickColor ?? DEFAULT_AXIS_FONT_COLOR
  };
  const defaultTickLabelProps = (...args) => ({
    ...fontProps,
    ...typeof tickLabelProps === "function" ? tickLabelProps(...args) : tickLabelProps
  });
  const truncatedLabel = label ? truncateString(
    label,
    containerSize,
    size === "small" ? charsFontWidth12pixelsSize400Map : charsFontWidth14pixelsSize400Map,
    "middle"
  ) : "";
  return /* @__PURE__ */ jsx(
    "svg",
    {
      className: axis,
      width,
      height,
      role: restProps.role,
      "aria-label": restProps["aria-label"] ?? void 0,
      children: /* @__PURE__ */ jsx(
        VisxAxis,
        {
          label: truncatedLabel,
          labelProps: size === "small" ? labelProps : labelBaseProps,
          tickLabelProps: defaultTickLabelProps,
          tickLength: AXIS_TICK_LENGTH,
          hideTicks: true,
          hideAxisLine: true,
          ...restProps
        }
      )
    }
  );
};
Axis["displayName"] = "Axis";
export {
  Axis
};
