// @dynatrace/strato-components/esm/charts/core/components/axes/constants.js
import Colors from "@dynatrace/strato-design-tokens/colors";
import Typography from "@dynatrace/strato-design-tokens/typography";
const AXIS_TICK_LENGTH = 0;
const X_AXES_TICK_LINE_OFFSET = 4;
const Y_AXES_TICK_LINE_OFFSET = 6;
const AXIS_LABEL_HEIGHT = 12;
const MIN_SCALE_STEPS = AXIS_LABEL_HEIGHT;
const FLEX_OFFSET = 4;
const X_AXES_LABEL_OFFSET = X_AXES_TICK_LINE_OFFSET + 6 - FLEX_OFFSET;
const Y_AXIS_LABEL_OFFSET = 6;
const TOP_AXIS_LINE_OFFSET = 18;
const TOP_AXIS_LABEL_OFFSET = 8;
const BOTTOM_AXIS_LINE_OFFSET = 8;
const MIN_AXIS_TRUNCATION_SIZE = 33;
const MAX_ANNOTATIONS_AREA_HEIGHT = 46;
const TICK_LINE_HEIGHT = 14;
const TICK_LINE_ICON_OFFSET = 4;
const Y_LABEL_AREA = AXIS_LABEL_HEIGHT + Y_AXIS_LABEL_OFFSET;
const X_LABEL_AREA = AXIS_LABEL_HEIGHT + X_AXES_LABEL_OFFSET;
const TICK_ALIGMENT = {
  Start: "start",
  Middle: "middle",
  End: "end"
};
const DEFAULT_AXIS_FONT_COLOR = Colors.Text.Neutral.Subdued;
const axisFontProps = {
  fill: DEFAULT_AXIS_FONT_COLOR,
  color: DEFAULT_AXIS_FONT_COLOR,
  fontSize: Typography.Text.Small.Emphasized.Size,
  fontFamily: Typography.Text.Small.Emphasized.Family
};
const axisFontBaseProps = {
  fill: DEFAULT_AXIS_FONT_COLOR,
  color: DEFAULT_AXIS_FONT_COLOR,
  fontSize: Typography.Text.Base.Emphasized.Size,
  fontFamily: Typography.Text.Base.Emphasized.Family
};
const EMOJI_REGEX = /(\p{Emoji_Presentation}|\p{Extended_Pictographic})/u;
export {
  AXIS_LABEL_HEIGHT,
  AXIS_TICK_LENGTH,
  BOTTOM_AXIS_LINE_OFFSET,
  DEFAULT_AXIS_FONT_COLOR,
  EMOJI_REGEX,
  MAX_ANNOTATIONS_AREA_HEIGHT,
  MIN_AXIS_TRUNCATION_SIZE,
  MIN_SCALE_STEPS,
  TICK_ALIGMENT,
  TICK_LINE_HEIGHT,
  TICK_LINE_ICON_OFFSET,
  TOP_AXIS_LABEL_OFFSET,
  TOP_AXIS_LINE_OFFSET,
  X_AXES_LABEL_OFFSET,
  X_AXES_TICK_LINE_OFFSET,
  X_LABEL_AREA,
  Y_AXES_TICK_LINE_OFFSET,
  Y_AXIS_LABEL_OFFSET,
  Y_LABEL_AREA,
  axisFontBaseProps,
  axisFontProps
};
