// @dynatrace/strato-components/editors/base-code-editor/configuration/getDefaultConfigurations.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var getDefaultConfigurations_exports = {};
__export(getDefaultConfigurations_exports, {
  getDefaultConfiguration: () => getDefaultConfiguration,
  getMultipleLinesDefaultConfiguration: () => getMultipleLinesDefaultConfiguration,
  getSingleLineDefaultConfiguration: () => getSingleLineDefaultConfiguration
});
module.exports = __toCommonJS(getDefaultConfigurations_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_autocomplete = require("@codemirror/autocomplete");
var import_commands = require("@codemirror/commands");
var import_language = require("@codemirror/language");
var import_lint = require("@codemirror/lint");
var import_search = require("@codemirror/search");
var import_state = require("@codemirror/state");
var import_view = require("@codemirror/view");
var import_highlight = require("@lezer/highlight");
var import_client = require("react-dom/client");
var import_borders = __toESM(require("@dynatrace/strato-design-tokens/borders"));
var import_box_shadows = __toESM(require("@dynatrace/strato-design-tokens/box-shadows"));
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
var import_spacings = __toESM(require("@dynatrace/strato-design-tokens/spacings"));
var import_typography = __toESM(require("@dynatrace/strato-design-tokens/typography"));
var import_strato_icons = require("@dynatrace/strato-icons");
var import_Text = require("../../../typography/text/Text.js");
var editorStyles = __toESM(require("../BaseCodeEditor.sty.js"));
var import_testIds = require("../testIds.js");
var import_render_node = require("../utils/render-node.js");
const Source = (source) => () => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { fontStyle: "code", textStyle: "small-emphasized", style: { width: 30 }, children: source });
const Type = (type) => () => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
  import_Text.Text,
  {
    fontStyle: "code",
    textStyle: "base",
    as: "span",
    className: editorStyles.autocompletionTypeText,
    children: type
  }
);
const getDefaultConfiguration = (intlProps, options = {
  required: false,
  size: "default"
}) => {
  const isSizeDefault = options.size === "default";
  return [
    (0, import_commands.history)(),
    import_state.EditorState.allowMultipleSelections.of(true),
    (0, import_language.bracketMatching)(),
    (0, import_autocomplete.closeBrackets)(),
    (0, import_autocomplete.autocompletion)({
      closeOnBlur: false,
      icons: false,
      optionClass: (completion) => {
        return completion.type ? `cm-completionOption-${completion.type}` : "";
      },
      addToOptions: [
        {
          position: 0,
          render: () => {
            const marker = document.createElement("span");
            marker.setAttribute("aria-hidden", "true");
            marker.style.display = "none";
            queueMicrotask(() => {
              marker.parentElement?.setAttribute("data-dt-name", "Suggestion");
            });
            return marker;
          }
        },
        {
          position: 21,
          render: (completion) => (0, import_render_node.render)(
            completion,
            typeof completion.prefix === "string" ? Source(completion.prefix) : completion.prefix,
            intlProps
          )
        },
        {
          position: 51,
          render: (completion) => (0, import_render_node.render)(
            completion,
            typeof completion.postfix === "string" ? Type(completion.postfix) : completion.postfix,
            intlProps
          )
        }
      ]
    }),
    (0, import_lint.linter)(() => []),
    // empty linter to enable theme for linter errors
    (0, import_language.syntaxHighlighting)(import_highlight.classHighlighter),
    (0, import_view.rectangularSelection)(),
    (0, import_view.highlightActiveLine)(),
    (0, import_view.highlightActiveLineGutter)(),
    (0, import_search.highlightSelectionMatches)(),
    import_view.EditorView.contentAttributes.of({
      [import_testIds.dataTestId]: import_testIds.baseCodeEditorContentTestId,
      "aria-haspopup": "listbox",
      "aria-required": options.required ? "true" : "false"
    }),
    import_view.EditorView.updateListener.of((update) => {
      if ((0, import_view.hasHoverTooltips)(update.state) && (0, import_autocomplete.completionStatus)(update.state) !== null) {
        update.view.dispatch({
          effects: import_view.closeHoverTooltips
        });
      }
    }),
    // By default the `DQLEditor` tooltip has `position: fixed` except for when the editor is placed
    // inside a transformed element. In that case, the tooltip has `position: absolute` and is cut off
    // in some scenarios due to the new stacking context created by the transformed element.
    // This extension ensures that the tooltip remains visible / is never cut off.
    (0, import_view.tooltips)({ parent: document.body }),
    import_view.EditorView.baseTheme({
      ".cm-tooltip": {
        background: `${import_colors.default.Background.Surface.Default} !important`,
        color: import_colors.default.Text.Neutral.Default,
        maxWidth: "500px",
        padding: import_spacings.default.Size4,
        borderRadius: import_borders.default.Radius.Surface.Subdued,
        boxShadow: import_box_shadows.default.Surface.Floating.Rest,
        border: "none !important"
      },
      ".cm-tooltip.cm-tooltip-autocomplete": {
        width: "500px"
      },
      ".cm-tooltip-below": {
        marginTop: ` ${import_spacings.default.Size4}`
      },
      ".cm-tooltip.cm-tooltip-autocomplete > ul > completion-section": {
        whiteSpace: "normal",
        textOverflow: "initial",
        wordBreak: "normal",
        overflowWrap: "anywhere"
      },
      '.cm-tooltip.cm-tooltip-autocomplete > ul > completion-section:has(+ li[role="option"].cm-empty-option)': {
        borderBottom: "0 !important"
      },
      '.cm-tooltip-autocomplete > ul > li[role="option"]': {
        display: "flex",
        flexWrap: "wrap",
        justifyContent: "flex-start",
        alignItems: "center",
        gap: import_spacings.default.Size4,
        padding: `${import_spacings.default.Size6} ${import_spacings.default.Size12} !important`
      },
      '.cm-tooltip-autocomplete > ul > li[role="option"].cm-empty-option': {
        visibility: "hidden",
        height: 0,
        overflow: "hidden",
        fontSize: 0,
        lineHeight: 0,
        padding: "0 !important",
        margin: 0
      },
      ".cm-completionLabel": {
        fontSize: isSizeDefault ? import_typography.default.Code.Base.Default.Size : import_typography.default.Code.Small.Default.Size,
        fontFamily: isSizeDefault ? import_typography.default.Code.Base.Default.Family : import_typography.default.Code.Small.Default.Family,
        color: import_colors.default.Text.Neutral.Default,
        fontWeight: import_typography.default.Code.Base.Default.Weight,
        lineHeight: import_typography.default.Code.Base.Default.LineHeight,
        textDecoration: import_typography.default.Code.Base.Default.TextDecoration,
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        textTransform: import_typography.default.Code.Base.Default.TextTransform,
        flexGrow: 3,
        overflow: "hidden",
        textOverflow: "ellipsis"
      },
      ".cm-tooltip.cm-tooltip-hover": {
        minWidth: "auto"
      },
      ".cm-tooltip.cm-tooltip-autocomplete > ul": {
        maxHeight: "20em !important"
      },
      ".cm-tooltip.cm-tooltip-autocomplete > ul > li": {
        position: "relative"
      },
      ".cm-tooltip.cm-tooltip-autocomplete > ul > li::after": {
        position: "absolute",
        /* Required or else the outline would be cut off by the listbox boundaries */
        inset: import_spacings.default.Size2,
        borderRadius: "inherit",
        content: " "
      },
      ".cm-tooltip.cm-tooltip-autocomplete li": {
        borderRadius: import_borders.default.Radius.Field.Default
      },
      ".cm-tooltip.cm-tooltip-autocomplete li[aria-selected]": {
        transition: "none",
        borderRadius: import_borders.default.Radius.Field.Default,
        boxShadow: `inset 0 0 0 ${import_borders.default.Width.Default} ${import_colors.default.Border.Neutral.Accent}`,
        backgroundColor: import_colors.default.Background.Field.Neutral.DefaultHover
      },
      ".cm-tooltip.cm-tooltip-autocomplete li[aria-selected]:focus-visible": {
        outline: "none"
      },
      ".cm-tooltip.cm-completionInfo": {
        padding: `calc(${import_spacings.default.Size4} + ${import_spacings.default.Size6}) ${import_spacings.default.Size16} !important`,
        marginLeft: import_spacings.default.Size4,
        marginRight: import_spacings.default.Size4,
        maxWidth: "500px",
        minWidth: "200px"
      },
      ".cm-tooltip-autocomplete > ul > li.cm-completionOption-action > div": {
        width: "100%"
      },
      ".cm-completionOption-action .cm-completionLabel": {
        display: "none"
      },
      ".cm-tooltip.cm-tooltip-autocomplete li:hover": {
        backgroundColor: `${import_colors.default.Background.Field.Neutral.DefaultHover}`
      }
    })
  ];
};
const getSingleLineDefaultConfiguration = (intlProps, options) => {
  return [
    import_state.EditorState.transactionFilter.of((tr) => tr.newDoc.lines > 1 ? [] : [tr]),
    ...getDefaultConfiguration(intlProps, options)
  ];
};
const getMultipleLinesDefaultConfiguration = (intlProps, options) => {
  const { required, gutterConfiguration, size } = options;
  return [
    (0, import_view.lineNumbers)(),
    (0, import_language.foldGutter)({
      markerDOM: (open) => {
        const element = document.createElement("div");
        element.style.display = "flex";
        element.style.alignItems = "center";
        element.style.marginTop = "2px";
        element.style.transform = "rotate(-90deg)";
        element.style.cursor = "pointer";
        element.style.minWidth = "16px";
        element.style.minHeight = "16px";
        if (open) {
          element.style.transform = "rotate(0deg)";
        }
        const root = (0, import_client.createRoot)(element);
        root.render(
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_strato_icons.ChevronDownSmallIcon,
            {
              size: "small",
              style: { width: import_spacings.default.Size16, height: import_spacings.default.Size16 }
            }
          )
        );
        return element;
      }
    }),
    (0, import_view.gutter)({
      lineMarker: (view, line) => {
        const length = view.state.doc.length;
        if (length === 0 || !gutterConfiguration) {
          return null;
        }
        const { lineNumber, marker } = gutterConfiguration;
        if (lineNumber < 1 || lineNumber > view.state.doc.lines) {
          return null;
        }
        const { from, to } = view.state.doc.line(lineNumber);
        return line.from === from && line.to === to ? new class extends import_view.GutterMarker {
          toDOM() {
            const element = document.createElement("div");
            const root = (0, import_client.createRoot)(element);
            root.render(marker);
            return element;
          }
        }() : null;
      }
    }),
    ...getDefaultConfiguration(intlProps, { required, size })
  ];
};
