// @dynatrace/strato-components/editors/base-code-editor/configuration/language-extensions.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var language_extensions_exports = {};
__export(language_extensions_exports, {
  getAutocompleteExtensions: () => getAutocompleteExtensions,
  getLanguageConfiguration: () => getLanguageConfiguration,
  getLanguageExtensions: () => getLanguageExtensions
});
module.exports = __toCommonJS(language_extensions_exports);
var import_autocomplete = require("@codemirror/autocomplete");
var import_language = require("@codemirror/language");
var import_autocompletion_extension = require("../auto-completion/autocompletion.extension.js");
var import_list_autocompletion_extension = require("../auto-completion/list/list-autocompletion-extension.js");
const emptyLanguage = import_language.StreamLanguage.define({
  token: (stream) => {
    stream.skipToEnd();
    return null;
  }
});
function getLanguageExtensions(language, providers, intlProps) {
  const configLanguage = language ?? emptyLanguage;
  return [
    configLanguage,
    ...getAutocompleteExtensions(providers || [], intlProps).map(
      (autocomplete) => configLanguage.data.of({ autocomplete })
    )
  ];
}
function getAutocompleteExtensions(providers, intlProps) {
  return providers.map(
    (p) => typeof p === "function" ? (0, import_autocompletion_extension.autocompletionExtension)(p, intlProps) : (0, import_list_autocompletion_extension.listAutocompletionExtension)(p, intlProps)
  );
}
function getLanguageConfiguration(languageConfig, intlProps) {
  let providers;
  if (languageConfig.autocompleteOptions) {
    providers = languageConfig.autocompleteOptions.providers;
  }
  return [
    getLanguageExtensions(languageConfig.language, providers, intlProps),
    (0, import_autocomplete.autocompletion)({
      activateOnTyping: languageConfig.autocompleteOptions?.activateOnTyping,
      optionClass: languageConfig.autocompleteOptions?.optionClass
    }),
    ...languageConfig.extendWith ?? []
  ];
}
