// @dynatrace/strato-geo/map/components/toolbar/constants.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var constants_exports = {};
__export(constants_exports, {
  TOOLBAR_MESSAGES: () => TOOLBAR_MESSAGES,
  initialA11yToolbarTabIndexContext: () => initialA11yToolbarTabIndexContext
});
module.exports = __toCommonJS(constants_exports);
var import_react_intl = require("react-intl");
const initialA11yToolbarTabIndexContext = [
  { "zoom-in": 0 },
  { "zoom-out": -1 },
  { reset: -1 },
  { "zoom-to-fit": -1 },
  { "download-data": -1 }
];
const TOOLBAR_MESSAGES = (0, import_react_intl.defineMessages)({
  zoomActionZoomIn: {
    id: "11eDVd4MkXWk8Bok",
    defaultMessage: "Zoom in",
    description: 'Text in a tooltip shown when hovering the "Zoom in" action in the geo map chart toolbar'
  },
  zoomActionZoomOut: {
    id: "5DTYqeFoXLMaXUHR",
    defaultMessage: "Zoom out",
    description: 'Text in a tooltip shown when hovering the "Zoom out" action in the geo map chart toolbar'
  },
  zoomActionReset: {
    id: "hVE7x0G0Q7/ih93A",
    defaultMessage: "Reset",
    description: 'Text in a tooltip shown when hovering the "Reset" action in the geo map chart toolbar'
  },
  zoomActionZoomToFit: {
    id: "EhcEZc1+Xc9wYsXg",
    defaultMessage: "Zoom to fit",
    description: 'Text in a tooltip shown when hovering the "Zoom to fit" action in the geo map chart toolbar'
  },
  downloadData: {
    id: "hlwWx47trlDMOy+V",
    defaultMessage: "Download Data as CSV",
    description: 'Text in a tooltip shown when hovering the "Download Data" action in the geo map chart toolbar'
  }
});
