// @dynatrace/strato-geo/map/components/toolbar/MapToolbar.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MapToolbar_exports = {};
__export(MapToolbar_exports, {
  MapToolbar: () => MapToolbar
});
module.exports = __toCommonJS(MapToolbar_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_maplibre = require("@vis.gl/react-maplibre");
var import_charts = require("@dynatrace/strato-components/charts");
var import_MapDownloadDataButton = require("./buttons/MapDownloadDataButton.js");
var import_MapResetButton = require("./buttons/MapResetButton.js");
var import_MapZoomInOutButtons = require("./buttons/MapZoomInOutButtons.js");
var import_MapZoomToFitButton = require("./buttons/MapZoomToFitButton.js");
var import_constants = require("./constants.js");
var import_use_map_config = require("../../hooks/use-map-config.js");
var import_use_map_raw_data = require("../../hooks/use-map-raw-data.js");
var import_map_initial_view_provider = require("../../providers/map-initial-view.provider.js");
var import_selectors = require("../../store/selectors.js");
var import_store = require("../../store/store.js");
var import_build_and_download_csv = require("../../utils/build-and-download-csv.js");
const defaultToolbarPlacement = "top-right";
const MapToolbar = () => {
  const { toolbar: toolbarConfig, interactions: interactionsConfig } = (0, import_use_map_config.useMapConfig)();
  const data = (0, import_use_map_raw_data.useMapRawData)();
  const { current: map } = (0, import_react_maplibre.useMap)();
  const { mouseInBounds } = (0, import_selectors.useMapOverlayState)();
  const setState = (0, import_store.useSetState)();
  const dispatch = (0, import_charts._useOverlayTooltipReducer)();
  const isZoomEnabled = toolbarConfig?.zoom?.enabled ?? interactionsConfig?.zoom?.enabled ?? false;
  const isZoomToFitEnabled = toolbarConfig?.zoom?.enabled ?? interactionsConfig?.zoomToFit?.enabled ?? false;
  const isDownloadDataEnabled = toolbarConfig?.downloadCSV?.enabled ?? false;
  const hasZoomControls = isZoomEnabled || isZoomToFitEnabled;
  const hasControls = hasZoomControls || isDownloadDataEnabled;
  const hasOnlyDownloadCSV = isDownloadDataEnabled && !hasZoomControls;
  const handleMouseEnter = () => {
    dispatch({ type: "RESET_TOOLTIP" })();
    if (map) {
      map.queryRenderedFeatures().forEach((feature) => {
        if (feature.id !== void 0 && feature.layer.source) {
          map.setFeatureState(
            { source: feature.layer.source, id: feature.id },
            { active: false, isAnyActive: false }
          );
        }
      });
    }
    setState((prev) => ({
      ...prev
    }));
  };
  const handleMouseLeave = () => {
    setState((prev) => ({
      ...prev,
      tooltip: { ...prev.tooltip, enabled: true }
    }));
  };
  const handleDownloadDataToolbar = () => {
    (0, import_build_and_download_csv.buildAndDownloadCsv)(data);
  };
  return hasControls ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_map_initial_view_provider.MapInitialViewProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_charts._ChartToolbarRenderer,
    {
      visible: mouseInBounds,
      collapsed: toolbarConfig?.collapsed,
      draggable: toolbarConfig?.draggable,
      placement: toolbarConfig?.placement ?? defaultToolbarPlacement,
      handleDownloadDataToolbar,
      isDownloadEnabled: !hasOnlyDownloadCSV && isDownloadDataEnabled,
      initialA11yState: import_constants.initialA11yToolbarTabIndexContext,
      onMouseEnter: handleMouseEnter,
      onMouseLeave: handleMouseLeave,
      menuDisplayMode: "never",
      children: [
        isZoomEnabled && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_MapZoomInOutButtons.MapZoomInOutButtons, {}),
        isZoomEnabled && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_MapResetButton.MapResetButton, {}),
        isZoomToFitEnabled && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_MapZoomToFitButton.MapZoomToFitButton, {}),
        isDownloadDataEnabled && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_MapDownloadDataButton.MapDownloadDataButton, {})
      ]
    }
  ) }) : null;
};
MapToolbar["displayName"] = "MapToolbar";
