// @dynatrace/strato-components/charts/categorical-bar/components/interactions/datapoints/CategoricalBarChartDatapoints.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CategoricalBarChartDatapoints_exports = {};
__export(CategoricalBarChartDatapoints_exports, {
  CategoricalBarChartDatapoints: () => CategoricalBarChartDatapoints
});
module.exports = __toCommonJS(CategoricalBarChartDatapoints_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_Canvas = require("../../../../core/components/canvas/Canvas.js");
var import_CanvasContext = require("../../../../core/components/canvas/CanvasContext.js");
var import_useGridPlotSize = require("../../../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_constants = require("../../../../core/components/data-point/constants.js");
var import_DataPointCanvas = require("../../../../core/components/data-point/DataPointCanvas.js");
var import_useCategoricalBarChartBaseConfig = require("../../../hooks/useCategoricalBarChartBaseConfig.js");
const CategoricalBarChartDatapoints = (props) => {
  const { width, height } = (0, import_useGridPlotSize.useGridPlotSize)();
  const { layout } = (0, import_useCategoricalBarChartBaseConfig.useCategoricalBarChartBaseConfig)();
  const { points, closestPoint } = props;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_CanvasContext.CanvasContext.Provider,
    {
      value: {
        shapeRefs: /* @__PURE__ */ new Map()
      },
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_Canvas.Canvas,
        {
          width: width + import_constants.OVERLAY_PADDING * 2,
          height: height + import_constants.OVERLAY_PADDING * 2,
          children: points && points.map((pointGroup, index) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_DataPointCanvas.DataPointCanvas,
            {
              valueAxis: layout === "vertical" ? "y" : "x",
              points: pointGroup,
              closestPoint
            },
            `pointGroup-${index}`
          ))
        }
      )
    }
  );
};
CategoricalBarChartDatapoints["displayName"] = "CategoricalBarChartDatapoints";
