// @dynatrace/strato-components/esm/charts/core/components/single-value-renderer/providers/font-size.provider.js
import { jsx } from "react/jsx-runtime";
import { useLayoutEffect, useRef } from "react";
import { useId } from "../../../../../core/hooks/useId.js";
import { useMergeRefs } from "../../../../../core/hooks/useMergeRefs.js";
import { useFontSizeForContainer } from "../../../hooks/useFontSizeForContainer.js";
import { useThrottledResizeObserver } from "../../../hooks/useThrottledResizeObserver.js";
import {
  DEFAULT_MAX_VALUE_FONT_SIZE,
  FONT_HEIGHT_RATIO,
  FONT_WIDTH_RATIO
} from "../constants.js";
import { FontSizeContext } from "../contexts/font-size.context.js";
import { useGlobalFontSize } from "../hooks/use-global-font-size.js";
const FontSizeProvider = ({
  text,
  suffix,
  children
}) => {
  const ref = useRef(null);
  const { refCallback, width = 0, height = 0 } = useThrottledResizeObserver();
  const mergedRef = useMergeRefs([refCallback, ref]);
  let preferredFontSize = useFontSizeForContainer({
    container: {
      width: width * FONT_WIDTH_RATIO,
      height: height * FONT_HEIGHT_RATIO
    },
    text,
    element: ref.current?.querySelector("span"),
    // element containing text font style
    suffix,
    suffixElement: ref.current?.querySelector("small")
    // element containing suffix font style
  });
  if (text.length === 0 || text === "00") {
    preferredFontSize = DEFAULT_MAX_VALUE_FONT_SIZE;
  }
  const id = useId();
  const { updateValueFontSize } = useGlobalFontSize();
  useLayoutEffect(() => {
    if (height && width) {
      updateValueFontSize?.(id, preferredFontSize);
    }
  }, [height, width, preferredFontSize]);
  return /* @__PURE__ */ jsx(
    FontSizeContext.Provider,
    {
      value: {
        fontSize: preferredFontSize,
        refCallback: mergedRef
      },
      children
    }
  );
};
export {
  FontSizeProvider
};
