// @dynatrace/strato-components/charts/categorical-bar/components/thresholds/CategoricalThresholdBar.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CategoricalThresholdBar_exports = {};
__export(CategoricalThresholdBar_exports, {
  CategoricalThresholdBar: () => CategoricalThresholdBar
});
module.exports = __toCommonJS(CategoricalThresholdBar_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_ThresholdBar = require("../../../core/components/threshold-bar/ThresholdBar.js");
var import_get_marker_pixel = require("../../../core/components/threshold-bar/utils/get-marker-pixel.js");
var import_get_marker_value = require("../../../core/components/threshold-bar/utils/get-marker-value.js");
var import_type_guards = require("../../../core/types/type-guards.js");
var import_selectors = require("../../hooks/store/selectors.js");
var import_useCategoricalBarChartBaseConfig = require("../../hooks/useCategoricalBarChartBaseConfig.js");
var import_useCategoricalBarChartScales = require("../../hooks/useCategoricalBarChartScales.js");
var import_useCategoricalBarChartThresholds = require("../../hooks/useCategoricalBarChartThresholds.js");
var import_useFormatter = require("../../hooks/useFormatter.js");
const CategoricalThresholdBar = ({
  length
}) => {
  const { layout } = (0, import_useCategoricalBarChartBaseConfig.useCategoricalBarChartBaseConfig)();
  const { valueScale } = (0, import_useCategoricalBarChartScales.useCategoricalBarChartScales)();
  const setState = (0, import_selectors.useSetState)();
  const formatter = (0, import_useFormatter.useFormatter)();
  const thresholds = (0, import_useCategoricalBarChartThresholds.useCategoricalBarChartThresholds)();
  const handleHover = (id) => {
    setState((prev) => ({
      ...prev,
      hoveredThresholdId: id
    }));
  };
  const getMarkerProps = (yFirst, yLast) => ({
    y0: (0, import_get_marker_pixel.getMarkerPixel)(yFirst, valueScale, layout),
    y1: (0, import_get_marker_pixel.getMarkerPixel)(yLast, valueScale, layout),
    value: (0, import_get_marker_value.getMarkerValue)({ y0: yFirst, y1: yLast }, formatter)
  });
  const markers = thresholds.map((threshold) => {
    const domainMin = (0, import_type_guards.isBoundedRange)(threshold.data) ? threshold.data.value : threshold.data.min ?? -Infinity;
    const domainMax = (0, import_type_guards.isBoundedRange)(threshold.data) ? threshold.data.value : threshold.data.max ?? Infinity;
    return {
      ...threshold,
      ...getMarkerProps(domainMin, domainMax)
    };
  });
  return markers.length ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ThresholdBar.ThresholdBar,
    {
      markers,
      length,
      layout,
      onHover: handleHover
    }
  ) : null;
};
CategoricalThresholdBar["displayName"] = "TimeseriesThresholdBar";
