// @dynatrace/strato-components/layouts/hooks/useMarginProps.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useMarginProps_exports = {};
__export(useMarginProps_exports, {
  useMarginProps: () => useMarginProps
});
module.exports = __toCommonJS(useMarginProps_exports);
var import_lodash_es = require("lodash");
var import_spacingToCss = require("../helpers/spacingToCss.js");
const defaultMargin = {
  margin: 0,
  marginTop: 0,
  marginRight: 0,
  marginBottom: 0,
  marginLeft: 0
};
const useMarginProps = (props, defaultMarginProps = defaultMargin) => {
  const {
    margin,
    m,
    marginX,
    mx,
    marginY,
    my,
    marginTop,
    mt,
    marginRight,
    mr,
    marginBottom,
    mb,
    marginLeft,
    ml
  } = props;
  const convertedDefaultMarginProps = {
    marginTop: (0, import_spacingToCss.spacingToCss)(
      defaultMarginProps.marginTop ?? defaultMarginProps.margin
    ),
    marginLeft: (0, import_spacingToCss.spacingToCss)(
      defaultMarginProps.marginLeft ?? defaultMarginProps.margin
    ),
    marginBottom: (0, import_spacingToCss.spacingToCss)(
      defaultMarginProps.marginBottom ?? defaultMarginProps.margin
    ),
    marginRight: (0, import_spacingToCss.spacingToCss)(
      defaultMarginProps.marginRight ?? defaultMarginProps.margin
    )
  };
  return (0, import_lodash_es.merge)({}, convertedDefaultMarginProps, {
    marginTop: (0, import_spacingToCss.spacingToCss)(mt ?? marginTop ?? my ?? marginY ?? m ?? margin),
    marginRight: (0, import_spacingToCss.spacingToCss)(
      mr ?? marginRight ?? mx ?? marginX ?? m ?? margin
    ),
    marginBottom: (0, import_spacingToCss.spacingToCss)(
      mb ?? marginBottom ?? my ?? marginY ?? m ?? margin
    ),
    marginLeft: (0, import_spacingToCss.spacingToCss)(ml ?? marginLeft ?? mx ?? marginX ?? m ?? margin)
  });
};
