// @dynatrace/strato-components/layouts/hooks/usePaddingProps.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var usePaddingProps_exports = {};
__export(usePaddingProps_exports, {
  usePaddingProps: () => usePaddingProps
});
module.exports = __toCommonJS(usePaddingProps_exports);
var import_lodash_es = require("lodash");
var import_spacingToCss = require("../helpers/spacingToCss.js");
const defaultPadding = {
  padding: 0,
  paddingTop: 0,
  paddingRight: 0,
  paddingBottom: 0,
  paddingLeft: 0
};
const usePaddingProps = (props, defaultPaddingProps = defaultPadding) => {
  const {
    padding,
    p,
    paddingX,
    px,
    paddingY,
    py,
    paddingTop,
    pt,
    paddingRight,
    pr,
    paddingBottom,
    pb,
    paddingLeft,
    pl
  } = props;
  const convertedDefaultPaddingProps = {
    paddingTop: (0, import_spacingToCss.spacingToCss)(
      defaultPaddingProps.paddingTop ?? defaultPaddingProps.padding
    ),
    paddingLeft: (0, import_spacingToCss.spacingToCss)(
      defaultPaddingProps.paddingLeft ?? defaultPaddingProps.padding
    ),
    paddingBottom: (0, import_spacingToCss.spacingToCss)(
      defaultPaddingProps.paddingBottom ?? defaultPaddingProps.padding
    ),
    paddingRight: (0, import_spacingToCss.spacingToCss)(
      defaultPaddingProps.paddingRight ?? defaultPaddingProps.padding
    )
  };
  return (0, import_lodash_es.merge)({}, convertedDefaultPaddingProps, {
    paddingTop: (0, import_spacingToCss.spacingToCss)(
      pt ?? paddingTop ?? py ?? paddingY ?? p ?? padding
    ),
    paddingRight: (0, import_spacingToCss.spacingToCss)(
      pr ?? paddingRight ?? px ?? paddingX ?? p ?? padding
    ),
    paddingBottom: (0, import_spacingToCss.spacingToCss)(
      pb ?? paddingBottom ?? py ?? paddingY ?? p ?? padding
    ),
    paddingLeft: (0, import_spacingToCss.spacingToCss)(
      pl ?? paddingLeft ?? px ?? paddingX ?? p ?? padding
    )
  });
};
