// @dynatrace/strato-components/esm/content/code-snippet/utils/highlighting-utils.js
const DEFAULT_MAX_TOKENIZATION_LENGTH = 1e5;
function countNewlines(str) {
  let count = 0;
  for (let i = 0; i < str.length; i++) {
    if (str.codePointAt(i) === 10) {
      count++;
    }
  }
  return count;
}
function buildLineNumbers(startLine, lineCount) {
  const numbers = new Array(lineCount);
  for (let i = 0; i < lineCount; i++) {
    numbers[i] = String(startLine + i);
  }
  return numbers.join("\n");
}
function splitCodeForHighlighting(code, maxLength = DEFAULT_MAX_TOKENIZATION_LENGTH) {
  if (code.length <= maxLength) {
    return {
      highlightCode: code,
      noHighlightInline: "",
      noHighlightCode: "",
      noHighlightLineNumbers: ""
    };
  }
  const highlight = code.substring(0, maxLength);
  const remaining = code.substring(maxLength);
  const firstNewline = remaining.indexOf("\n");
  const inlinePart = firstNewline < 0 ? remaining : remaining.substring(0, firstNewline);
  const blockPart = firstNewline < 0 ? "" : remaining.substring(firstNewline + 1);
  if (blockPart.length === 0) {
    return {
      highlightCode: highlight,
      noHighlightInline: inlinePart,
      noHighlightCode: "",
      noHighlightLineNumbers: ""
    };
  }
  const highlightedLineCount = countNewlines(highlight) + 1;
  const startLine = highlightedLineCount + 1;
  const blockLineCount = countNewlines(blockPart) + 1;
  return {
    highlightCode: highlight,
    noHighlightInline: inlinePart,
    noHighlightCode: blockPart,
    noHighlightLineNumbers: buildLineNumbers(startLine, blockLineCount)
  };
}
export {
  splitCodeForHighlighting
};
