// @dynatrace/strato-components/content/code-snippet/CodeHighlight.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CodeHighlight_exports = {};
__export(CodeHighlight_exports, {
  CodeHighlight: () => CodeHighlight
});
module.exports = __toCommonJS(CodeHighlight_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_error_boundary = require("react-error-boundary");
var styles = __toESM(require("./CodeHighlight.sty.js"));
var import_HighlightExtended = require("./HighlightExtended.js");
var import_highlighting_utils = require("./utils/highlighting-utils.js");
var import_sdk_logger = require("../../core/sdk/sdk-logger.js");
function NonHighlightedBlock({
  showLineNumbers,
  size,
  isMultiLine,
  lineNumbers,
  code
}) {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { className: styles.line, children: [
    showLineNumbers ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "span",
      {
        className: (0, import_clsx.default)(
          "line-number",
          styles.lineNumberContainer({ multiLine: isMultiLine })
        ),
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          "span",
          {
            className: styles.lineNumber({
              size,
              multiLine: isMultiLine
            }),
            children: lineNumbers
          }
        )
      }
    ) : null,
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "span",
      {
        className: styles.lineContent({
          size,
          multiLine: isMultiLine
        }),
        children: code
      }
    )
  ] });
}
const CodeHighlight = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    code,
    language,
    showLineNumbers = true,
    lineBreaks = false,
    maxHeight,
    size = "default"
  } = props;
  const stratoLogger = (0, import_sdk_logger.getStratoLogger)("CodeSnippet");
  const {
    highlightCode,
    noHighlightInline,
    noHighlightCode,
    noHighlightLineNumbers
  } = (0, import_react.useMemo)(() => (0, import_highlighting_utils.splitCodeForHighlighting)(code), [code]);
  const isNonHighlightedMultiLine = noHighlightLineNumbers.includes("\n");
  const hasNonHighlightedBlock = noHighlightCode.length > 0;
  const isFullyHighlighted = noHighlightInline.length === 0 && !hasNonHighlightedBlock;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "pre",
    {
      className: styles.highlightWrapper({ size }),
      ref: forwardedRef,
      style: {
        whiteSpace: lineBreaks ? "pre-wrap" : "pre",
        maxHeight
      },
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_react_error_boundary.ErrorBoundary,
        {
          fallback: code,
          onError: (error, info) => {
            stratoLogger.error(error.message, error, {
              code,
              info
            });
          },
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_HighlightExtended.HighlightExtended, { language, code: highlightCode, children: ({ tokens, getLineProps, getTokenProps }) => {
            const isSingleLine = tokens.length === 1 && isFullyHighlighted;
            return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
              tokens.map((line, i) => {
                const isLastLine = i === tokens.length - 1;
                const { className: lineClassName, ...remainingLineProps } = getLineProps({ line });
                return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
                  "div",
                  {
                    className: (0, import_clsx.default)(lineClassName, styles.line),
                    ...remainingLineProps,
                    children: [
                      showLineNumbers ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                        "span",
                        {
                          className: (0, import_clsx.default)(
                            "line-number",
                            styles.lineNumberContainer()
                          ),
                          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                            "span",
                            {
                              className: styles.lineNumber({ size }),
                              "data-linenumber": i + 1
                            }
                          )
                        }
                      ) : null,
                      /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
                        "span",
                        {
                          className: styles.lineContent({ size }),
                          style: {
                            verticalAlign: isSingleLine ? "middle" : void 0
                          },
                          children: [
                            line.map((token, index) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { ...getTokenProps({ token }) }, index)),
                            isLastLine && noHighlightInline.length > 0 ? noHighlightInline : null
                          ]
                        }
                      )
                    ]
                  },
                  i
                );
              }),
              hasNonHighlightedBlock ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                NonHighlightedBlock,
                {
                  showLineNumbers,
                  size,
                  isMultiLine: isNonHighlightedMultiLine,
                  lineNumbers: noHighlightLineNumbers,
                  code: noHighlightCode
                }
              ) : null
            ] });
          } })
        }
      )
    }
  );
});
CodeHighlight.displayName = "CodeHighlight";
