// @dynatrace/strato-components/content/code-snippet/CodeSnippet.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CodeSnippet_exports = {};
__export(CodeSnippet_exports, {
  CodeSnippet: () => CodeSnippet
});
module.exports = __toCommonJS(CodeSnippet_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_copy_to_clipboard = __toESM(require("copy-to-clipboard"));
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_borders = __toESM(require("@dynatrace/strato-design-tokens/borders"));
var import_strato_icons = require("@dynatrace/strato-icons");
var import_CodeHighlight = require("./CodeHighlight.js");
var styles = __toESM(require("./CodeSnippet.sty.js"));
var import_Button = require("../../buttons/button/Button.js");
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useSafeForwardProps = require("../../core/hooks/useSafeForwardProps.js");
var import_Tooltip = require("../../overlays/tooltip/Tooltip.js");
const COMPONENT_NAME = "CodeSnippet";
const CodeSnippet = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    children,
    language,
    // eslint-disable-next-line @typescript-eslint/no-deprecated
    allowCopy: allowCopyProp,
    showCopyAction: showCopyActionProp,
    showLineNumbers = true,
    lineBreaks = false,
    maxHeight,
    onCopy,
    size = "default",
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    style: consumerStyle,
    className: consumerClassName,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const ariaLabelingProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(restProps);
  const forwardProps = (0, import_useSafeForwardProps.useSafeForwardProps)(restProps, ariaLabelingProps);
  const showCopyAction = showCopyActionProp ?? allowCopyProp ?? true;
  const joinedChildren = (0, import_react.useMemo)(
    () => Array.isArray(children) ? children.join("") : children,
    [children]
  );
  const copyText = `${joinedChildren?.toString() ?? ""}`;
  const [hasCopied, setHasCopied] = (0, import_react.useState)(false);
  const [isTooltipOpen, setIsTooltipOpen] = (0, import_react.useState)(false);
  const intl = (0, import_react_intl.useIntl)();
  const handleCopy = async () => {
    (0, import_copy_to_clipboard.default)(copyText, { format: "text/plain" });
    setHasCopied(true);
    setIsTooltipOpen(true);
    onCopy?.();
  };
  const copySelection = (0, import_react.useCallback)(
    (event) => {
      const selection = document.getSelection();
      if (selection === null) {
        return;
      }
      event.clipboardData.setData("text/plain", selection.toString());
      event.preventDefault();
    },
    []
  );
  const copyLabel = intl.formatMessage({
    defaultMessage: "Copy",
    id: "Yi8j4dchtpkwn/OB",
    description: "Shown for elements that trigger a copy to clipboard"
  });
  const copiedLabel = intl.formatMessage({
    defaultMessage: "Copied",
    id: "THNwSvfvDSXAUapi",
    description: "Shown after copying to clipboard"
  });
  const message = (0, import_react.useMemo)(() => {
    return hasCopied ? copiedLabel : copyLabel;
  }, [hasCopied, copiedLabel, copyLabel]);
  const codeHighlightRef = (0, import_react.useRef)(null);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    "div",
    {
      ref: forwardedRef,
      onCopy: copySelection,
      className: (0, import_clsx.default)(consumerClassName, styles.wrapper),
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...ariaLabelingProps,
      ...behavioralTrackingProps,
      ...forwardProps,
      "data-dt-component": COMPONENT_NAME,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_CodeHighlight.CodeHighlight,
          {
            ref: codeHighlightRef,
            language,
            code: copyText,
            showLineNumbers,
            lineBreaks,
            maxHeight,
            size
          }
        ),
        showCopyAction ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: styles.copyButtonContainer({ size }), children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_Tooltip.Tooltip,
          {
            text: message,
            placement: "top",
            open: isTooltipOpen,
            delay: "none",
            onOpenChange: (isOpen) => {
              setIsTooltipOpen(isOpen);
              if (!isOpen) {
                setHasCopied(false);
              }
            },
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_Button.Button,
              {
                variant: "default",
                size,
                "aria-label": copyLabel,
                onClick: handleCopy,
                style: { borderRadius: import_borders.default.Radius.Field.Subdued },
                "data-dt-name": "Copy",
                children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.CopyIcon, {}) })
              }
            )
          }
        ) }) : null
      ]
    }
  );
});
CodeSnippet.displayName = COMPONENT_NAME;
