// @dynatrace/strato-components/content/code-snippet/HighlightExtended.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var HighlightExtended_exports = {};
__export(HighlightExtended_exports, {
  HighlightExtended: () => HighlightExtended
});
module.exports = __toCommonJS(HighlightExtended_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_prism_react_renderer = require("prism-react-renderer");
var import_react = require("react");
var import_prism_dql_language = require("./prism-dql-language.js");
var import_logging = require("../../core/utils/logging.js");
var import_prism_theme = require("../../core/utils/prism-theme.js");
(typeof global !== "undefined" ? global : window).Prism = import_prism_react_renderer.Prism;
import_prism_react_renderer.Prism.languages.dql = import_prism_dql_language.prismDqlLanguage;
const SHORT_HEX_ID_TOKEN_NAME = "strato-short-hash";
const SHORT_HEX_ID_PATTERN = /\b[a-f0-9]{16}\b/i;
function enhancePrismLogLanguage() {
  const logLanguage = import_prism_react_renderer.Prism.languages.log;
  if (!logLanguage || SHORT_HEX_ID_TOKEN_NAME in logLanguage) {
    return;
  }
  import_prism_react_renderer.Prism.languages.insertBefore("log", "property", {
    [SHORT_HEX_ID_TOKEN_NAME]: {
      pattern: SHORT_HEX_ID_PATTERN,
      alias: "constant"
    }
  });
}
const additionalLanguageImports = {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  bash: () => Promise.resolve().then(() => __toESM(require("prismjs/components/prism-bash.js"))),
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  log: () => Promise.resolve().then(() => __toESM(require("prismjs/components/prism-log.js"))),
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  groovy: () => Promise.resolve().then(() => __toESM(require("prismjs/components/prism-groovy.js")))
};
const HighlightExtended = (props) => {
  const { language, code, children } = props;
  const [loadedLanguages, setLoadedLanguages] = (0, import_react.useState)(
    () => Object.fromEntries(
      Object.keys(additionalLanguageImports).map((lang) => [
        lang,
        import_prism_react_renderer.Prism.languages[lang] !== void 0
      ])
    )
  );
  (0, import_react.useEffect)(() => {
    if (language && additionalLanguageImports[language] && !loadedLanguages[language]) {
      additionalLanguageImports[language]().then(() => {
        if (language === "log") {
          enhancePrismLogLanguage();
        }
        setLoadedLanguages((prev) => ({ ...prev, [language]: true }));
      }).catch((error) => {
        (0, import_logging.devLog)(`Failed to load language '${language}':`, error);
      });
    }
  }, [language, loadedLanguages]);
  (0, import_react.useEffect)(() => {
    if (language === "log" && loadedLanguages.log) {
      enhancePrismLogLanguage();
    }
  }, [language, loadedLanguages.log]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_prism_react_renderer.Highlight,
    {
      language: language ?? "plain",
      code,
      theme: language === "log" ? import_prism_theme.DTPrismTheme.Log : import_prism_theme.DTPrismTheme.Default,
      Prism: import_prism_react_renderer.Prism,
      children: ({ tokens, getLineProps, getTokenProps, className, style }) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, { children: children({
        tokens,
        getLineProps,
        getTokenProps,
        className,
        style
      }) })
    }
  );
};
