// @dynatrace/strato-components/content/code-snippet/prism-dql-language.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var prism_dql_language_exports = {};
__export(prism_dql_language_exports, {
  prismDqlLanguage: () => prismDqlLanguage
});
module.exports = __toCommonJS(prism_dql_language_exports);
var import_dql_common = require("../../core/dql/dql-common.js");
const singleLineCommentPattern = /\/\/[^\n\r\v\f]*/;
const multiLineCommentPattern = /\/\*[\s\S]*?\*\//m;
const comments = {
  comment: {
    pattern: singleLineCommentPattern,
    greedy: true
  },
  commentMultiline: {
    pattern: multiLineCommentPattern,
    alias: "comment",
    greedy: true
  }
};
const stringPattern = /"(?:[^"\\]|\\.)*"/m;
const backtickStringPattern = /`(?:[^`\\]|\\.)*`/m;
const tripleQuoteStringPattern = /""""?"?(?:[^"]"?"?)*"""/m;
const dqlCommandNamePattern = new RegExp(
  `\\b(?:${import_dql_common.sortedDqlCommands.join("|")})\\b`,
  "i"
);
const commandNamePattern = /[A-Za-z]+/;
const commandParametersPattern = new RegExp(
  `(${commandNamePattern.source})(?:${tripleQuoteStringPattern.source}|${backtickStringPattern.source}|${stringPattern.source}|${singleLineCommentPattern.source}|${multiLineCommentPattern.source}|[^|])*`
);
const commandPattern = new RegExp(
  `${commandNamePattern.source}${commandParametersPattern.source}\\|?`,
  "m"
);
const commandName = {
  commandName: {
    pattern: dqlCommandNamePattern,
    alias: "keyword"
  }
};
const prismDqlLanguage = {
  ...comments,
  /**
   * DQL language is defined as commands separated by the optional character (|).
   * Prism first matches an entire command, then matches the command parameters and command name
   * 'inside' the command.
   * Within the command parameters, Prism matches strings, variables, and numbers.
   *
   * Take great care of adding or modifying the regex patterns as this
   * may introduce recursion issues, due to the existence of three levels of regex matching.
   */
  command: {
    pattern: commandPattern,
    greedy: true,
    inside: {
      commandParameters: {
        pattern: commandParametersPattern,
        lookbehind: true,
        inside: {
          tripleQuoteString: {
            pattern: tripleQuoteStringPattern,
            alias: "string"
          },
          string: {
            pattern: stringPattern
            // Support strings surrounded by double quotes and include possible escaped quotes within
          },
          backtickstring: {
            pattern: backtickStringPattern,
            alias: "string"
          },
          number: new RegExp(`\\b\\d+(?:\\.\\d+)?${import_dql_common.dqlUnitSuffixPattern}?\\b`),
          // Match numbers with optional unit suffix
          boolean: import_dql_common.booleanLiteralPattern,
          // Match boolean literals (true/false)
          null: { pattern: import_dql_common.nullLiteralPattern, alias: "boolean" },
          // null literal, styled as boolean
          parameterName: { pattern: import_dql_common.parameterNamePattern, alias: "variable" },
          ...comments,
          commandName: {
            pattern: new RegExp(
              `(\\[\\s*)${dqlCommandNamePattern.source}`,
              "i"
            ),
            lookbehind: true,
            alias: "keyword"
          }
        },
        ...commandName
      },
      ...commandName
    }
  }
};
