// @dynatrace/strato-components/esm/tables/DataTable/contexts/DataTableContext.js
import { jsx } from "react/jsx-runtime";
import {
  useContext,
  createContext
} from "react";
const DataTableContext = createContext(null);
function DataTableContextProvider(props) {
  return /* @__PURE__ */ jsx(DataTableContext.Provider, { value: props.value, children: props.children });
}
function useDataTable() {
  const table = useContext(DataTableContext);
  if (table === null) {
    throw new Error("Table instance is not defined in context.");
  }
  return table;
}
const DEFAULT_CONTENT_CONTEXT = {
  hasOverflowLeft: false,
  hasOverflowRight: false,
  cellActionAriaLabel: ""
};
const DataTableContentContext = createContext(DEFAULT_CONTENT_CONTEXT);
function DataTableContentContextProvider(props) {
  return /* @__PURE__ */ jsx(
    DataTableContentContext.Provider,
    {
      value: props.value ?? DEFAULT_CONTENT_CONTEXT,
      children: props.children
    }
  );
}
function useDataTableContent() {
  const content = useContext(
    DataTableContentContext
  );
  if (content === null) {
    throw new Error("Content instance is not defined in context.");
  }
  return content;
}
const DataTableHeaderContext = createContext(null);
function DataTableHeaderContextProvider(props) {
  return /* @__PURE__ */ jsx(DataTableHeaderContext.Provider, { value: props.value ?? null, children: props.children });
}
function useDataTableHeader() {
  const header = useContext(
    DataTableHeaderContext
  );
  if (header === null) {
    throw new Error("Header instance is not defined in context.");
  }
  return header;
}
const DataTableColumnContext = createContext(
  null
);
function DataTableColumnContextProvider(props) {
  return /* @__PURE__ */ jsx(DataTableColumnContext.Provider, { value: props.value ?? null, children: props.children });
}
function useDataTableColumn() {
  const column = useContext(
    DataTableColumnContext
  );
  if (column === null) {
    throw new Error("Column object is not defined in context.");
  }
  return column;
}
const DataTableRowContext = createContext(null);
function DataTableRowContextProvider(props) {
  return /* @__PURE__ */ jsx(DataTableRowContext.Provider, { value: props.value ?? null, children: props.children });
}
function useDataTableRow() {
  const row = useContext(DataTableRowContext);
  return row;
}
const DataTableCellContext = createContext(null);
function DataTableCellContextProvider(props) {
  return /* @__PURE__ */ jsx(DataTableCellContext.Provider, { value: props.value ?? null, children: props.children });
}
function useDataTableCell() {
  const cell = useContext(DataTableCellContext);
  return cell;
}
const DataTableColumnSettingsContext = createContext(null);
function DataTableColumnSettingsContextProvider(props) {
  return /* @__PURE__ */ jsx(DataTableColumnSettingsContext.Provider, { value: props.value ?? null, children: props.children });
}
function useDataTableColumnSettings() {
  return useContext(
    DataTableColumnSettingsContext
  );
}
const DataTableIsHeaderContext = createContext(false);
function DataTableIsHeaderContextProvider(props) {
  return /* @__PURE__ */ jsx(DataTableIsHeaderContext.Provider, { value: true, children: props.children });
}
function useDataTableIsHeaderContextProvider() {
  return useContext(DataTableIsHeaderContext);
}
export {
  DataTableCellContextProvider,
  DataTableColumnContextProvider,
  DataTableColumnSettingsContextProvider,
  DataTableContentContext,
  DataTableContentContextProvider,
  DataTableContextProvider,
  DataTableHeaderContext,
  DataTableHeaderContextProvider,
  DataTableIsHeaderContextProvider,
  DataTableRowContextProvider,
  useDataTable,
  useDataTableCell,
  useDataTableColumn,
  useDataTableColumnSettings,
  useDataTableContent,
  useDataTableHeader,
  useDataTableIsHeaderContextProvider,
  useDataTableRow
};
