// @dynatrace/strato-components/esm/charts/honeycomb/utils/iterate-honeycomb-slots.js
import { Children } from "react";
import { Intent } from "../../../core/slots/Intent/Intent.js";
import { isSlot } from "../../../core/utils/is-slot.js";
import { ColorRule, isValidColorRule } from "../../core/slots/ColorRule.js";
import { ChartToolbar } from "../../core/slots/toolbar/ChartToolbar.js";
import { sanitizeProps } from "../../core/slots/toolbar/CompactToolbar.js";
import { DownloadData } from "../../core/slots/toolbar/DownloadData.js";
import { extractBehaviorTrackingProps } from "../../core/utils/extract-behavior-tracking-props.js";
import { DEFAULT_LEGEND_POSITION } from "../constants.js";
import { HoneycombChartDownloadCSVSlot } from "../slots/DownloadCSV.js";
import { HoneycombLegend } from "../slots/HoneycombLegend.js";
import { Toolbar } from "../slots/Toolbar.js";
import { Tooltip } from "../slots/Tooltip.js";
const buildChartToolbarConfigSlot = (child, config) => {
  const { children, ...toolbar } = child.props;
  const childrenToolbar = iterateChartToolbarSlots(children);
  return {
    ...config,
    toolbar: {
      ...toolbar,
      ...childrenToolbar
    }
  };
};
const iterateHoneycombSlots = (children) => {
  let config = {
    legend: void 0,
    toolbar: void 0,
    colorRules: [],
    tooltip: void 0
  };
  let intents;
  Children.forEach(children, (child) => {
    if (isSlot(child, HoneycombLegend)) {
      config = {
        ...config,
        legend: {
          position: DEFAULT_LEGEND_POSITION,
          ...child.props
        }
      };
    } else if (isSlot(child, ChartToolbar)) {
      config = buildChartToolbarConfigSlot(child, config);
    } else if (isSlot(child, Toolbar)) {
      config.toolbar = {
        ...config.toolbar,
        ...sanitizeProps(child.props, ["draggable", "hidden", "placement"]),
        ...extractBehaviorTrackingProps(child.props)
      };
    } else if (isSlot(child, HoneycombChartDownloadCSVSlot)) {
      config = {
        ...config,
        toolbar: { ...config.toolbar, downloadCSV: { enabled: true } }
      };
    } else if (isSlot(child, Intent)) {
      intents = intents ? [...intents, child.props] : [child.props];
    } else if (isSlot(child, ColorRule)) {
      isValidColorRule(child.props) && config.colorRules.push(child.props);
    } else if (isSlot(child, Tooltip)) {
      config = {
        ...config,
        tooltip: {
          hidden: child.props?.hidden ?? false,
          template: child.props?.children
        }
      };
    }
  });
  if (intents) {
    config = {
      ...config,
      toolbar: {
        ...config.toolbar,
        intents,
        hidden: config.toolbar?.hidden ?? false
      }
    };
  }
  return config;
};
const iterateChartToolbarSlots = (children) => {
  let toolbar = {};
  Children.forEach(children, (childToolbar) => {
    if (isSlot(childToolbar, DownloadData)) {
      toolbar = {
        ...toolbar,
        downloadCSV: { enabled: true }
      };
    }
  });
  return toolbar;
};
export {
  iterateHoneycombSlots
};
