// @dynatrace/strato-components/charts/xy-chart/components/toolbar/DownloadDataButton.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DownloadDataButton_exports = {};
__export(DownloadDataButton_exports, {
  DownloadDataButton: () => DownloadDataButton
});
module.exports = __toCommonJS(DownloadDataButton_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_ChartToolbarRenderer = require("../../../core/components/toolbar/ChartToolbarRenderer.js");
var import_useA11yToolbarTabIndex = require("../../../core/hooks/useA11yToolbarTabIndex.js");
var import_useDownload = require("../../hooks/useDownload.js");
var import_useXYChartData = require("../../hooks/useXYChartData.js");
var import_messages = require("../../messages/messages.js");
var import_build_and_download_csv = require("../../utils/build-and-download-csv.js");
const DownloadDataButton = () => {
  const downloadDataButtonRef = (0, import_react.useRef)(null);
  const tabIndex = (0, import_useA11yToolbarTabIndex.useA11yToolbarTabIndex)("download");
  const intl = (0, import_react_intl.useIntl)();
  const { downloadDataCallback } = (0, import_useDownload.useDownload)();
  const { seriesWithDatapoints, datapointMetadataMap, seriesToMetadata } = (0, import_useXYChartData.useXYChartData)();
  const triggerDownload = () => {
    (0, import_build_and_download_csv.buildAndDownloadCsv)(
      seriesWithDatapoints,
      seriesToMetadata,
      datapointMetadataMap,
      downloadDataCallback
    );
  };
  const handleDownloadDataKeyboard = (event) => {
    if (event.key === "Enter") {
      triggerDownload();
    }
  };
  const handleDownloadDataToolbar = () => {
    triggerDownload();
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ChartToolbarRenderer.ChartToolbarRenderer.Control,
    {
      ref: downloadDataButtonRef,
      prefixIcon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.DownloadIcon, {}),
      tooltipText: intl.formatMessage(import_messages.XYCHART_MESSAGES.downloadAsCsv),
      onKeyDown: handleDownloadDataKeyboard,
      onClick: handleDownloadDataToolbar,
      tabIndex,
      divider: "left",
      "data-dt-name": "ToolbarDownloadData"
    }
  );
};
DownloadDataButton["displayName"] = "DownloadDataButton";
