// @dynatrace/strato-components/charts/xy-chart/components/toolbar/PanButton.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var PanButton_exports = {};
__export(PanButton_exports, {
  PanButton: () => PanButton
});
module.exports = __toCommonJS(PanButton_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_utils = require("./utils.js");
var import_constants = require("../../../core/components/interactions-overlay/constants.js");
var import_types = require("../../../core/components/interactions-overlay/types.js");
var import_useSynchronizedSelection = require("../../../core/components/shared-chart-interactions/hooks/useSynchronizedSelection.js");
var import_ChartToolbarRenderer = require("../../../core/components/toolbar/ChartToolbarRenderer.js");
var import_messages = require("../../../core/components/toolbar/messages.js");
var import_useA11yToolbarTabIndex = require("../../../core/hooks/useA11yToolbarTabIndex.js");
var import_get_key_shortcut_by_agent = require("../../../core/utils/get-key-shortcut-by-agent.js");
var import_messages2 = require("../../messages/messages.js");
var import_reduce_interaction_mode = require("../../store/reduce-interaction-mode.js");
var import_reducer_explorer = require("../../store/reducer-explorer.js");
var import_selectors = require("../../store/selectors.js");
const PanButton = () => {
  const setState = (0, import_selectors.useSetState)();
  const panButtonRef = (0, import_react.useRef)(null);
  const intl = (0, import_react_intl.useIntl)();
  const tabIndex = (0, import_useA11yToolbarTabIndex.useA11yToolbarTabIndex)(import_types.INTERACTION_MODE.Pan);
  const { zoomOutEnabled } = (0, import_selectors.useZoomState)();
  const interactionMode = (0, import_selectors.useInteractionModeState)();
  const { resetSelectionState } = (0, import_useSynchronizedSelection.useSynchronizedSelection)();
  const isSelected = interactionMode === import_types.INTERACTION_MODE.Pan;
  const action = () => {
    resetSelectionState?.();
    setState(import_reducer_explorer.reduceExploreCancelAll);
    setState((0, import_reduce_interaction_mode.reduceInteractionMode)(import_types.INTERACTION_MODE.Pan));
  };
  const tooltipKeyShortcut = `${(0, import_get_key_shortcut_by_agent.getKeyShortcutByAgent)(
    "ALT"
  )} ${intl.formatMessage(import_messages.TOOLBAR_MESSAGES.or)} ${import_constants.KeyboardControls.KEYBOARD_PAN_MODE_SHORTCUT_ALTERNATIVE}`;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ChartToolbarRenderer.ChartToolbarRenderer.Control,
    {
      ref: panButtonRef,
      prefixIcon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.PanHorizontalIcon, {}),
      tooltipText: intl.formatMessage(import_messages2.XYCHART_MESSAGES.zoomModePan),
      tooltipKeyShortcut,
      selected: isSelected,
      onKeyDown: (0, import_utils.allowWhenKeyIsEnter)(action),
      onClick: action,
      tabIndex,
      disabled: !zoomOutEnabled,
      placement: "menu",
      "data-dt-name": "ToolbarPanMode"
    }
  );
};
PanButton["displayName"] = "PanButton";
