// @dynatrace/strato-components/charts/xy-chart/components/toolbar/ResetButton.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ResetButton_exports = {};
__export(ResetButton_exports, {
  ResetButton: () => ResetButton
});
module.exports = __toCommonJS(ResetButton_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_utils = require("./utils.js");
var import_constants = require("../../../core/components/interactions-overlay/constants.js");
var import_ChartToolbarRenderer = require("../../../core/components/toolbar/ChartToolbarRenderer.js");
var import_useA11yToolbarTabIndex = require("../../../core/hooks/useA11yToolbarTabIndex.js");
var import_zoom = require("../../../core/types/zoom.js");
var import_useXYChartData = require("../../hooks/useXYChartData.js");
var import_useXYChartScales = require("../../hooks/useXYChartScales.js");
var import_messages = require("../../messages/messages.js");
var import_useResetInteractions = require("../overlay/hooks/useResetInteractions.js");
const ResetButton = () => {
  const tabIndex = (0, import_useA11yToolbarTabIndex.useA11yToolbarTabIndex)(import_zoom.ZOOM_CHANGE_TYPE.Reset);
  const { axesMetadata, axes } = (0, import_useXYChartData.useXYChartData)();
  const axisIdToScales = (0, import_useXYChartScales.useXYChartScales)();
  const intl = (0, import_react_intl.useIntl)();
  const action = (0, import_useResetInteractions.useResetInteractions)();
  const resetDisabled = axesMetadata.isZoomEnabled && (0, import_utils.areAllDomainsUnchanged)(axes, axisIdToScales);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ChartToolbarRenderer.ChartToolbarRenderer.Control,
    {
      prefixIcon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ResetIcon, {}),
      tooltipText: intl.formatMessage(import_messages.XYCHART_MESSAGES.resetMode),
      tooltipKeyShortcut: import_constants.KeyboardControls.KEYBOARD_RESET_SHORTCUT_ALTERNATIVE,
      onClick: action,
      onKeyDown: (0, import_utils.allowWhenKeyIsEnter)(action),
      tabIndex,
      disabled: resetDisabled,
      placement: "toolbar",
      showWhenCollapsed: !resetDisabled,
      "data-dt-name": "ToolbarReset"
    }
  );
};
ResetButton["displayName"] = "ResetButton";
