// @dynatrace/strato-components/charts/xy-chart/components/toolbar/XYChartToolbar.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var XYChartToolbar_exports = {};
__export(XYChartToolbar_exports, {
  XYChartToolbar: () => XYChartToolbar
});
module.exports = __toCommonJS(XYChartToolbar_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_intl = require("react-intl");
var import_DownloadDataButton = require("./DownloadDataButton.js");
var import_ExploreButton = require("./ExploreButton.js");
var import_PanButton = require("./PanButton.js");
var import_ResetButton = require("./ResetButton.js");
var import_ZoomInOutButtons = require("./ZoomInOutButtons.js");
var import_ZoomModeButton = require("./ZoomModeButton.js");
var import_useBehavioralTrackingProps = require("../../../../core/hooks/useBehavioralTrackingProps.js");
var import_ChartToolbarRenderer = require("../../../core/components/toolbar/ChartToolbarRenderer.js");
var import_get_menu_trigger_display_mode = require("../../../core/components/toolbar/utils/get-menu-trigger-display-mode.js");
var import_get_menu_trigger_icon = require("../../../core/components/toolbar/utils/get-menu-trigger-icon.js");
var import_constants = require("../../../core/constants.js");
var import_zoom = require("../../../core/utils/zoom.js");
var import_defaults = require("../../constants/defaults.js");
var import_useDownload = require("../../hooks/useDownload.js");
var import_useToolbar = require("../../hooks/useToolbar.js");
var import_useXYChartData = require("../../hooks/useXYChartData.js");
var import_useXYChartSelectionAreaRef = require("../../hooks/useXYChartSelectionAreaRef.js");
var import_messages = require("../../messages/messages.js");
var import_selectors = require("../../store/selectors.js");
var import_build_and_download_csv = require("../../utils/build-and-download-csv.js");
const XYChartToolbar = ({
  onMouseEnter,
  onMouseLeave,
  handlerRef
}) => {
  const toolbarContext = (0, import_useToolbar.useToolbar)();
  const { collapsed, placement, hidden, intents } = toolbarContext;
  const { downloadDataCallback } = (0, import_useDownload.useDownload)();
  const interactionMode = (0, import_selectors.useInteractionModeState)();
  const { axesMetadata } = (0, import_useXYChartData.useXYChartData)();
  const intl = (0, import_react_intl.useIntl)();
  const selectionAreaRef = (0, import_useXYChartSelectionAreaRef.useXYChartSelectionAreaRef)();
  const isZoomEnabled = axesMetadata.isZoomEnabled;
  const isPanEnabled = axesMetadata.isPanEnabled;
  const exploreBtnPlace = isZoomEnabled || isPanEnabled ? "menu" : "auto";
  const isDownloadEnabled = true;
  const initialA11yState = (0, import_zoom.getA11yToolbarConfig)({
    zoom: { enabled: isZoomEnabled },
    pan: { enabled: isPanEnabled },
    download: { enabled: isDownloadEnabled },
    intent: !!intents
  });
  const menuIcon = (0, import_get_menu_trigger_icon.getMenuTriggerIcon)(interactionMode, isZoomEnabled);
  const menuLabel = !isPanEnabled && !isZoomEnabled ? intl.formatMessage(import_messages.XYCHART_MESSAGES.zoomModeExplore) : void 0;
  const menuDisplayMode = (0, import_get_menu_trigger_display_mode.getMenuTriggerDisplayMode)({
    isPanEnabled,
    isZoomEnabled,
    isDownloadEnabled
  });
  const { seriesWithDatapoints, datapointMetadataMap, seriesToMetadata } = (0, import_useXYChartData.useXYChartData)();
  const handleDownloadDataToolbar = () => {
    (0, import_build_and_download_csv.buildAndDownloadCsv)(
      seriesWithDatapoints,
      seriesToMetadata,
      datapointMetadataMap,
      downloadDataCallback
    );
  };
  const showDefaultToolbarDownloadButton = isDownloadEnabled && !intents;
  const showMenuToolbarButton = isDownloadEnabled && !!intents?.length;
  const onClose = () => selectionAreaRef ? selectionAreaRef.current?.focus() : void 0;
  const behaviorTrackingProps = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(toolbarContext)[0];
  return !hidden ? /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_ChartToolbarRenderer.ChartToolbarRenderer,
    {
      visible: false,
      visibleWhenMouseInBounds: true,
      handlerRef,
      placement: placement ?? import_defaults.defaultToolbar.placement,
      draggable: true,
      collapsed,
      initialA11yState,
      zIndex: import_constants.ChartOverlaysElevation.toolbar,
      onMouseEnter,
      onMouseLeave,
      menuIcon,
      menuTriggerLabel: menuLabel,
      menuDisplayMode,
      isDownloadEnabled: showMenuToolbarButton,
      intents,
      handleDownloadDataToolbar,
      onClose,
      ...behaviorTrackingProps,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ExploreButton.ExploreButton, { placement: exploreBtnPlace }),
          isZoomEnabled && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ZoomModeButton.ZoomModeButton, {}),
          isPanEnabled && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_PanButton.PanButton, {})
        ] }),
        isZoomEnabled && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ZoomInOutButtons.ZoomInOutButtons, {}),
        isZoomEnabled && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ResetButton.ResetButton, {}),
        showDefaultToolbarDownloadButton && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_DownloadDataButton.DownloadDataButton, {})
      ]
    }
  ) : null;
};
XYChartToolbar["displayName"] = "XYChartToolbar";
