// @dynatrace/strato-components/charts/xy-chart/components/toolbar/ZoomModeButton.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ZoomModeButton_exports = {};
__export(ZoomModeButton_exports, {
  ZoomModeButton: () => ZoomModeButton
});
module.exports = __toCommonJS(ZoomModeButton_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_useGridPlotSize = require("../../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_constants = require("../../../core/components/interactions-overlay/constants.js");
var import_types = require("../../../core/components/interactions-overlay/types.js");
var import_calculate_initial_selection_area_positions = require("../../../core/components/interactions-overlay/utils/calculate-initial-selection-area-positions.js");
var import_mode_reducers = require("../../../core/components/interactions-state-helpers/reducers/mode-reducers.js");
var import_OverlayContext = require("../../../core/components/overlay-tooltip/contexts/OverlayContext.js");
var import_useSynchronizedSelection = require("../../../core/components/shared-chart-interactions/hooks/useSynchronizedSelection.js");
var import_ChartToolbarRenderer = require("../../../core/components/toolbar/ChartToolbarRenderer.js");
var import_messages = require("../../../core/components/toolbar/messages.js");
var import_useA11yToolbarTabIndex = require("../../../core/hooks/useA11yToolbarTabIndex.js");
var import_get_key_shortcut_by_agent = require("../../../core/utils/get-key-shortcut-by-agent.js");
var import_get_percentage_from_pixel = require("../../../core/utils/get-percentage-from-pixel.js");
var import_messages2 = require("../../messages/messages.js");
var import_reduce_interaction_mode = require("../../store/reduce-interaction-mode.js");
var import_reducer_explorer = require("../../store/reducer-explorer.js");
var import_selectors = require("../../store/selectors.js");
const ZoomModeButton = () => {
  const setState = (0, import_selectors.useSetState)();
  const interactionMode = (0, import_selectors.useInteractionModeState)();
  const tabIndex = (0, import_useA11yToolbarTabIndex.useA11yToolbarTabIndex)(import_types.INTERACTION_MODE.Zoom);
  const { overlayRef } = (0, import_OverlayContext.useOverlay)();
  const intl = (0, import_react_intl.useIntl)();
  const selectState = (0, import_selectors.useSelectState)();
  const { width } = (0, import_useGridPlotSize.useGridPlotSize)();
  const isSelected = interactionMode === import_types.INTERACTION_MODE.Zoom;
  const { resetSelectionState } = (0, import_useSynchronizedSelection.useSynchronizedSelection)();
  const handleZoomMode = () => {
    if (interactionMode !== "zoom") {
      setState((0, import_reduce_interaction_mode.reduceInteractionMode)(import_types.INTERACTION_MODE.Zoom));
    }
    if (selectState?.start && selectState?.end) {
      if (selectState?.isKeyboardCreated) {
        (0, import_mode_reducers.reduceSelectionAreaVariant)("zoom");
      } else {
        resetSelectionState?.();
        setState(import_reducer_explorer.reduceExploreCancelAll);
      }
    }
    setState((0, import_reduce_interaction_mode.reduceInteractionMode)(import_types.INTERACTION_MODE.Zoom));
  };
  const handleKeyDown = (event) => {
    if (event.key === "Enter") {
      if (!(0, import_lodash_es.isNil)(overlayRef.current)) {
        const bounds = overlayRef.current.getBoundingClientRect();
        const [start, end] = (0, import_calculate_initial_selection_area_positions.calculateInitialSelectionAreaPositions)(bounds);
        if (interactionMode !== import_types.INTERACTION_MODE.Explore) {
          setState((0, import_reduce_interaction_mode.reduceInteractionMode)(import_types.INTERACTION_MODE.Explore));
        }
        setState(
          (0, import_reducer_explorer.reduceExploreSelect)({
            start,
            end,
            finished: true,
            accessible: true,
            variant: "zoom",
            selectionPercentage: (0, import_get_percentage_from_pixel.getPercentageFromPixel)([start, end], width)
          })
        );
      }
    }
  };
  const tooltipKeyShortcut = `${(0, import_get_key_shortcut_by_agent.getKeyShortcutByAgent)(
    "CONTROL"
  )} ${intl.formatMessage(import_messages.TOOLBAR_MESSAGES.or)} ${import_constants.KeyboardControls.KEYBOARD_ZOOM_MODE_SHORTCUT_ALTERNATIVE}`;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ChartToolbarRenderer.ChartToolbarRenderer.Control,
    {
      prefixIcon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ZoomToSelectionIcon, {}),
      selected: isSelected,
      tooltipText: intl.formatMessage(import_messages2.XYCHART_MESSAGES.zoomModeZoom),
      onClick: handleZoomMode,
      onKeyDown: handleKeyDown,
      tabIndex,
      tooltipKeyShortcut,
      placement: "menu",
      "data-dt-name": "ToolbarZoomMode"
    }
  );
};
