// @dynatrace/strato-components/charts/core/color-rule/strategies.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var strategies_exports = {};
__export(strategies_exports, {
  strategies: () => strategies
});
module.exports = __toCommonJS(strategies_exports);
var import_lodash_es = require("lodash");
var import_utils = require("./utils.js");
var import_getAccessor = require("../utils/getAccessor.js");
var import_getValidColor = require("../utils/getValidColor.js");
var import_validators = require("../utils/validators.js");
const strategies = {
  histogram: () => {
    return {
      defaultValueAccessor: "name",
      forbiddenAccessors: /^name$/,
      getName: (series) => series.name,
      getSubject: import_utils.getSubject,
      matchers: import_validators.VALUE_MATCHERS,
      colorPaletteIndex: import_utils.incBy1,
      getColorFromRule: import_utils.getColorFromRule
    };
  },
  timeseries: () => {
    return {
      defaultValueAccessor: "name",
      forbiddenAccessors: /^name$/,
      getName: (series) => Array.isArray(series.name) ? series.name[0] : series.name,
      getSubject: import_utils.getSubject,
      matchers: import_validators.VALUE_MATCHERS,
      colorPaletteIndex: import_utils.incBy1,
      getColorFromRule: import_utils.getColorFromRule
    };
  },
  categoricalBar: () => {
    return {
      defaultValueAccessor: "category",
      forbiddenAccessors: /^category$/,
      getName: (series) => series.category,
      getSubject: import_utils.getSubject,
      matchers: import_validators.VALUE_MATCHERS,
      colorPaletteIndex: import_utils.always0,
      getColorFromRule: (rule) => {
        if (rule.colorPalette) {
          return rule.colorPalette;
        }
        return { Default: (0, import_getValidColor.getValidColor)(rule.color ?? "") };
      }
    };
  },
  meterBar: () => {
    const getValueWithAccessor = (currentSeries, valueAccessor) => {
      const accessor = (0, import_getAccessor.getAccessor)(valueAccessor);
      if (!accessor) {
        return null;
      }
      const value = (0, import_lodash_es.get)(currentSeries, accessor);
      return !(0, import_lodash_es.isNil)(value) ? value : null;
    };
    return {
      defaultValueAccessor: "value",
      forbiddenAccessors: /^value$/,
      getName: (series) => JSON.stringify(series.value),
      getSubject: function(rule, currentSeries) {
        const valueAccessor = rule.valueAccessor || currentSeries["_chartValueAccessor"];
        if (valueAccessor) {
          return getValueWithAccessor(currentSeries, valueAccessor);
        }
        return currentSeries.value;
      },
      matchers: import_validators.VALUE_MATCHERS,
      colorPaletteIndex: import_utils.incBy1,
      getColorFromRule: import_utils.getColorFromRule
    };
  },
  pieOrDonut: () => {
    return {
      defaultValueAccessor: "value",
      forbiddenAccessors: /^(value|relativeValue)$/,
      getName: (series) => series.category,
      getSubject: import_utils.getSubject,
      matchers: import_validators.VALUE_MATCHERS,
      colorPaletteIndex: import_utils.incBy1,
      getColorFromRule: import_utils.getColorFromRule
    };
  },
  honeycomb: () => {
    return {
      defaultValueAccessor: "value",
      forbiddenAccessors: /^value$/,
      getName: () => "honeycomb",
      getSubject: import_utils.getSubject,
      matchers: import_validators.VALUE_MATCHERS,
      colorPaletteIndex: import_utils.always0,
      getColorFromRule: import_utils.getColorFromRule
    };
  },
  categoricalLegend: () => {
    return {
      defaultValueAccessor: "valueOf",
      forbiddenAccessors: /(?:)/,
      getName: () => "categoricalLegend",
      getSubject: (_rule, item) => item,
      matchers: import_validators.VALUE_MATCHERS,
      colorPaletteIndex: import_utils.always0,
      getColorFromRule: import_utils.getColorFromRule
    };
  },
  numericalLegend: () => {
    return {
      defaultValueAccessor: "from",
      forbiddenAccessors: /(?:)/,
      getName: () => "numericalLegend",
      getSubject: (_, item) => item,
      matchers: import_validators.RANGE_MATCHERS,
      colorPaletteIndex: import_utils.always0,
      getColorFromRule: import_utils.getColorFromRule
    };
  }
};
