// @dynatrace/strato-components/charts/xy-chart/processing-strategies/utils/gap/get-gap-policy-information.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var get_gap_policy_information_exports = {};
__export(get_gap_policy_information_exports, {
  detectGapBetweenDatapoints: () => detectGapBetweenDatapoints,
  getAxisGapPolicyThreshold: () => getAxisGapPolicyThreshold,
  getXDatapointDimensionValues: () => getXDatapointDimensionValues,
  getYDatapointDimensionValues: () => getYDatapointDimensionValues
});
module.exports = __toCommonJS(get_gap_policy_information_exports);
var import_gap_policy = require("../../../../core/types/gap-policy.js");
var import_apply_gap_policy = require("../../../../core/utils/apply-gap-policy.js");
var import_get_data_point_is_gap = require("../../../../core/utils/get-data-point-is-gap.js");
var import_guards = require("../../../types/guards.js");
function getAxisGapPolicyThreshold(axis) {
  if ((0, import_guards.isCategoricalAxis)(axis)) {
    return Infinity;
  }
  if ((0, import_guards.isTimeAxis)(axis)) {
    return axis.gapPolicy === import_gap_policy.GAP_POLICY_CONNECT ? Infinity : (0, import_apply_gap_policy.getThresholdValue)(axis.gapPolicy ?? import_gap_policy.GAP_POLICY_GAP);
  }
  return axis.gapPolicy === import_gap_policy.GAP_POLICY_CONNECT || axis.gapPolicy === void 0 ? Infinity : (0, import_apply_gap_policy.getNumericalThresholdValue)(axis.gapPolicy);
}
function getDatapointDimensionValues(datapoint, axis, accessorKeys) {
  if ((0, import_guards.isTimeAxis)(axis)) {
    return {
      start: datapoint[accessorKeys[0]],
      end: datapoint[accessorKeys[1]]
    };
  }
  return {
    start: datapoint[accessorKeys[0]] ?? datapoint[accessorKeys[1]]
  };
}
function getXDatapointDimensionValues(datapoint, axis) {
  return getDatapointDimensionValues(datapoint, axis, ["x0", "x1"]);
}
function getYDatapointDimensionValues(datapoint, axis) {
  return getDatapointDimensionValues(datapoint, axis, ["y0", "y1"]);
}
function detectGapBetweenDatapoints(lastDatapoint, currentDatapoint, xAxis, yAxis) {
  const xGapPolicyThreshold = getAxisGapPolicyThreshold(xAxis);
  const yGapPolicyThreshold = getAxisGapPolicyThreshold(yAxis);
  let isGap = false;
  if (xGapPolicyThreshold !== Infinity) {
    const lastDatapointX = getXDatapointDimensionValues(lastDatapoint, xAxis);
    const currentDatapointX = getXDatapointDimensionValues(
      currentDatapoint,
      xAxis
    );
    isGap = (0, import_get_data_point_is_gap.getDatapointIsGap)(
      lastDatapointX,
      currentDatapointX,
      xGapPolicyThreshold
    );
  }
  if (!isGap && yGapPolicyThreshold !== Infinity) {
    const lastDatapointY = getYDatapointDimensionValues(lastDatapoint, yAxis);
    const currentDatapointY = getYDatapointDimensionValues(
      currentDatapoint,
      yAxis
    );
    isGap = (0, import_get_data_point_is_gap.getDatapointIsGap)(
      lastDatapointY,
      currentDatapointY,
      yGapPolicyThreshold
    );
  }
  return isGap;
}
