// @dynatrace/strato-components/esm/forms/search-input/SearchInput.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import { forwardRef } from "react";
import { MagnifyingGlassIcon } from "@dynatrace/strato-icons";
import * as styles from "./SearchInput.sty.js";
import { Stepper } from "./Stepper.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useControllableState } from "../../core/hooks/useControllableState.js";
import { useSlots } from "../../core/hooks/useSlots.js";
import { mergeProps } from "../../core/utils/merge-props.js";
import { BaseInput } from "../_base-input/BaseInput.js";
import { BaseInputButton } from "../_base-input/BaseInputButton.js";
import { BaseInputSuffix } from "../_base-input/BaseInputContent.js";
import { ClearButton } from "../_base-input/ClearButton.js";
const COMPONENT_NAME = "SearchInput";
const SearchInputRoot = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const {
    children,
    onChange,
    value,
    defaultValue,
    placeholder,
    required,
    readOnly = false,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "aria-disabled": ariaDisabledProp,
    className: consumerClassName,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
  const [internalValue, setInternalValue] = useControllableState({
    defaultValue,
    value,
    onChange
  });
  const handleOnChange = (e) => {
    setInternalValue(e.target.value);
    if (slots.stepper && e.target.value === "") {
      const { onCurrentChange } = slots.stepper.props;
      onCurrentChange(0);
    }
  };
  const handleKeyDown = (e) => {
    if (!slots.stepper) {
      return;
    }
    if (e.key !== "Enter") {
      return;
    }
    e.preventDefault();
    const { current, total, onCurrentChange } = slots.stepper.props;
    if (e.shiftKey) {
      onCurrentChange(current > 0 ? current - 1 : total - 1);
    } else {
      onCurrentChange(current < total - 1 ? current + 1 : 0);
    }
  };
  function onClear() {
    setInternalValue("");
  }
  const [slots] = useSlots(children, {
    stepper: Stepper
  });
  const isEmpty = internalValue === "" || internalValue === void 0;
  const hasStepper = !!slots.stepper;
  return /* @__PURE__ */ jsxs(
    BaseInput,
    {
      ref: forwardedRef,
      ...mergeProps(restProps, behavioralTrackingProps),
      ...hasStepper ? { onKeyDown: handleKeyDown } : {},
      className: clsx(styles.searchInput({ hasStepper }), consumerClassName),
      type: "search",
      value: internalValue,
      required,
      readOnly,
      onChange: handleOnChange,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "aria-disabled": ariaDisabledProp,
      placeholder,
      "data-dt-component": COMPONENT_NAME,
      size: 1,
      children: [
        /* @__PURE__ */ jsx(BaseInput.Prefix, { children: /* @__PURE__ */ jsx(MagnifyingGlassIcon, {}) }),
        !isEmpty && /* @__PURE__ */ jsxs(BaseInput.ReservedSuffix, { children: [
          slots.stepper,
          /* @__PURE__ */ jsx(ClearButton, { onClear })
        ] }),
        children
      ]
    }
  );
});
const SearchInput = Object.assign(SearchInputRoot, {
  Suffix: BaseInputSuffix,
  Button: BaseInputButton,
  Stepper
});
SearchInput.displayName = COMPONENT_NAME;
export {
  SearchInput
};
