// @dynatrace/strato-components/esm/forms/search-input/Stepper.js
import { jsx, jsxs } from "react/jsx-runtime";
import { useContext } from "react";
import { useIntl } from "react-intl";
import { ChevronDownIcon, ChevronUpIcon } from "@dynatrace/strato-icons";
import * as styles from "./Stepper.sty.js";
import { Button } from "../../buttons/button/Button.js";
import { devLog } from "../../core/utils/logging.js";
import { Divider } from "../../layouts/divider/Divider.js";
import { Flex } from "../../layouts/flex/Flex.js";
import * as baseInputStyles from "../_base-input/BaseInput.sty.js";
import { BaseInputContext } from "../_base-input/contexts/BaseInputContext.js";
const Stepper = ({
  current,
  onCurrentChange,
  total
}) => {
  const intl = useIntl();
  const { hasError, disabled, ariaDisabled } = useContext(BaseInputContext);
  const disabledOrAriaDisabled = disabled || ariaDisabled || false;
  if (current > total) {
    devLog(
      `The current match index (${current}) is greater than the total number of matches (${total}). Please ensure that the current index is zero-based and less than the total.`
    );
  }
  return /* @__PURE__ */ jsxs(Flex, { gap: 4, children: [
    /* @__PURE__ */ jsxs(
      "div",
      {
        className: styles.matchPositionIndicators,
        role: "status",
        "aria-live": "polite",
        "aria-label": intl.formatMessage(
          {
            id: "tw9ulrSgegYXHq69",
            defaultMessage: "Match {current} of {total}",
            description: "Label for the current match position"
          },
          { current: current + 1, total }
        ),
        children: [
          /* @__PURE__ */ jsx("span", { children: current + 1 }),
          " / ",
          /* @__PURE__ */ jsx("span", { children: total })
        ]
      }
    ),
    /* @__PURE__ */ jsx(Divider, { orientation: "vertical", className: baseInputStyles.divider }),
    /* @__PURE__ */ jsxs("div", { className: styles.stepperButtons, role: "group", children: [
      /* @__PURE__ */ jsx(
        Button,
        {
          onClick: (e) => {
            e.stopPropagation();
            onCurrentChange(current === 0 ? total - 1 : current - 1);
          },
          className: styles.stepperButton,
          color: hasError ? "critical" : "neutral",
          disabled: disabledOrAriaDisabled || total === 0,
          "aria-label": intl.formatMessage({
            id: "GLK6cN59tiRO9GDp",
            defaultMessage: "Previous match",
            description: "Label for the previous match button"
          }),
          children: /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(ChevronUpIcon, {}) })
        }
      ),
      /* @__PURE__ */ jsx(
        Button,
        {
          onClick: (e) => {
            e.stopPropagation();
            onCurrentChange(current === total - 1 ? 0 : current + 1);
          },
          className: styles.stepperButton,
          color: hasError ? "critical" : "neutral",
          disabled: disabledOrAriaDisabled || total === 0,
          "aria-label": intl.formatMessage({
            id: "JLMJNMOPi2dCAbD1",
            defaultMessage: "Next match",
            description: "Label for the next match button"
          }),
          children: /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(ChevronDownIcon, {}) })
        }
      )
    ] })
  ] });
};
Stepper.displayName = "SearchInput.Stepper";
export {
  Stepper
};
