// @dynatrace/strato-components/charts/gauge/components/GaugeLayoutOverlay/GaugeLayoutOverlay.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var GaugeLayoutOverlay_exports = {};
__export(GaugeLayoutOverlay_exports, {
  GaugeChartOverlay: () => GaugeChartOverlay
});
module.exports = __toCommonJS(GaugeLayoutOverlay_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_GaugeLayoutOverlay_css = require("./GaugeLayoutOverlay.sty.js");
var import_Flex = require("../../../../layouts/flex/Flex.js");
var import_LoaderOverlay = require("../../../core/components/loader-overlay/LoaderOverlay.js");
var import_OverlayContext = require("../../../core/components/overlay-tooltip/contexts/OverlayContext.js");
var import_useTooltipActions = require("../../../core/components/overlay-tooltip/useTooltipActions.js");
var import_GaugeChartTooltipContext = require("../../contexts/GaugeChartTooltipContext.js");
var import_useGaugeChartTooltipOverlayBounds = require("../../hooks/useGaugeChartTooltipOverlayBounds.js");
var import_GaugeChartToolbar = require("../GaugeChartLayout/GaugeChartToolbar.js");
var import_GaugeOverlayTooltip = require("../GaugeChartTooltip/GaugeOverlayTooltip.js");
const GaugeChartOverlay = (props) => {
  const { children, value, color, intents, toolbar, showLoading } = props;
  const overlay = (0, import_OverlayContext.useOverlay)();
  const overlayBounds = (0, import_useGaugeChartTooltipOverlayBounds.useGaugeChartTooltipOverlayBounds)();
  const { name: tooltipName, hidden: tooltipHidden } = (0, import_GaugeChartTooltipContext.useGaugeChartTooltip)();
  const { hover, togglePin, outOfBounds } = (0, import_useTooltipActions.useTooltipActions)();
  const data = {
    geometry: "meterbar",
    series: [{ name: tooltipName ?? "Value", value, color }]
  };
  const overlayClassName = overlayBounds ? import_GaugeLayoutOverlay_css.tooltipOverlay : `${import_GaugeLayoutOverlay_css.tooltipOverlay} ${import_GaugeLayoutOverlay_css.tooltipOverlayFull}`;
  const overlayStyle = overlayBounds ? {
    top: overlayBounds.top,
    left: overlayBounds.left,
    width: overlayBounds.width,
    height: overlayBounds.height
  } : void 0;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_Flex.Flex,
    {
      ref: overlay.overlayRef,
      "data-testid": "gauge-overlay-wrapper",
      className: (0, import_GaugeLayoutOverlay_css.container)({ showLoading }),
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_GaugeChartToolbar.GaugeChartToolbar, { intents, toolbar }),
        !showLoading && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_Flex.Flex,
          {
            className: overlayClassName,
            onPointerMove: (event) => hover(event, data),
            onPointerLeave: outOfBounds,
            onClick: (event) => togglePin(event, data),
            style: overlayStyle,
            children: !tooltipHidden && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_GaugeOverlayTooltip.GaugeOverlayTooltip, { value, color })
          }
        ),
        children,
        showLoading ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_LoaderOverlay.LoaderOverlay,
          {
            gridArea: "1/1/2/2",
            consumerStyle: import_GaugeLayoutOverlay_css.consumerLoadingStyle
          }
        ) : null
      ]
    }
  );
};
GaugeChartOverlay["displayName"] = "GaugeChartOverlay";
