// @dynatrace/strato-components/charts/tree-map/components/TreeMapOverlay/TreeMapOverlay.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TreeMapOverlay_exports = {};
__export(TreeMapOverlay_exports, {
  TreeMapOverlay: () => TreeMapOverlay
});
module.exports = __toCommonJS(TreeMapOverlay_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_TreeMapOverlay_css = require("./TreeMapOverlay.sty.js");
var import_Flex = require("../../../../layouts/flex/Flex.js");
var import_useGraphSize = require("../../../core/components/chart-layout/hooks/useGraphSize.js");
var import_utils = require("../../../core/components/overlay-tooltip/_to-be-removed/utils.js");
var import_OverlayContext = require("../../../core/components/overlay-tooltip/contexts/OverlayContext.js");
var import_ChartSeriesActionContext = require("../../../core/components/series-action/ChartSeriesActionContext.js");
var import_is_toolbar_enabled = require("../../../core/utils/is-toolbar-enabled.js");
var import_ToolbarContext = require("../../context/ToolbarContext.js");
var import_TreeMapOverlayRefContext = require("../../contexts/TreeMapOverlayRefContext.js");
var import_useOverlayHandlers = require("../../hooks/useOverlayHandlers.js");
var import_TreeMapToolbar = require("../TreeMapToolbar/TreeMapToolbar.js");
var import_TreeMapOverlayTooltip = require("../TreeMapTooltip/TreeMapOverlayTooltip.js");
const TreeMapOverlay = () => {
  const overlay = (0, import_OverlayContext.useOverlay)();
  const { width, height } = (0, import_useGraphSize.useGraphSize)();
  const roundedWidth = Math.floor(width);
  const roundedHeight = Math.floor(height);
  const {
    handlePointerEnter,
    handlePointerMove,
    handlePointerLeave,
    handleMouseClick
  } = (0, import_useOverlayHandlers.useOverlayHandlers)(overlay);
  const toolbarConfig = (0, import_ToolbarContext.useToolbarConfig)();
  const toolbarEnabled = (0, import_is_toolbar_enabled.isToolbarEnabled)(toolbarConfig);
  const { chartSeriesActions } = (0, import_ChartSeriesActionContext.useSeriesActions)();
  const data = (0, import_utils.useTooltipData)().data;
  const series = data?.series?.[0];
  const metadata = data?.metadata;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TreeMapOverlayRefContext.TreeMapOverlayRefContext.Provider, { value: overlay.overlayRef, children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_Flex.Flex,
    {
      ref: overlay.overlayRef,
      style: { width: roundedWidth, height: roundedHeight },
      className: import_TreeMapOverlay_css.container,
      onClick: handleMouseClick,
      onPointerEnter: handlePointerEnter,
      onPointerMove: handlePointerMove,
      onPointerLeave: handlePointerLeave,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_TreeMapOverlayTooltip.TreeMapOverlayTooltip,
          {
            actionsForItem: chartSeriesActions && series && metadata ? () => () => chartSeriesActions({
              id: metadata.id,
              name: series.name,
              value: metadata.value
            }) : void 0
          }
        ),
        toolbarEnabled && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TreeMapToolbar.TreeMapToolbar, {})
      ]
    }
  ) });
};
TreeMapOverlay["displayName"] = "TreeMapOverlay";
