// @dynatrace/strato-components/tables/DataTable/features/RowOrder/row-order-model.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var row_order_model_exports = {};
__export(row_order_model_exports, {
  getRowOrderRowModel: () => getRowOrderRowModel
});
module.exports = __toCommonJS(row_order_model_exports);
var import_is_node_environment = require("../../../../core/utils/is-node-environment.js");
var import_constants = require("../../constants.js");
var import_src = require("../../tanstack-table-core-override/src/index.js");
function getRowOrderRowModel() {
  return (table) => {
    return (0, import_src.memo)(
      () => [
        table.getPreRowOrderRowModel(),
        table.options.enableRowOrdering,
        table.getState().rowOrder,
        table.options.state.sorting
      ],
      (rowModel) => {
        const rowOrder = table.getState().rowOrder;
        if (table.options.enableRowOrdering !== true || table.options.state.sorting && table.options.state.sorting.length !== 0) {
          return rowModel;
        }
        const { flatRows, rows, rowsById } = rowModel;
        if (rowOrder === void 0) {
          table.setRowOrder(flatRows.map((row) => row.id));
          return rowModel;
        }
        const orderRows = (rows2) => {
          const orderedData = [...rows2];
          orderedData.sort(
            (rowA, rowB) => rowOrderCompareFn(rowOrder, rowA, rowB)
          );
          orderedData.forEach((row) => {
            if (row.subRows?.length) {
              row.subRows = orderRows(row.subRows);
            }
          });
          return orderedData;
        };
        const newRowsById = Object.keys(rowsById).sort((a, b) => rowOrderCompareStrFn(rowOrder, a, b)).reduce((obj, key) => {
          obj[key] = rowsById[key];
          return obj;
        }, {});
        return {
          flatRows: orderRows(flatRows),
          rows: orderRows(rows),
          rowsById: newRowsById
        };
      },
      {
        key: (0, import_is_node_environment.isDev)() && "_getRowOrderRowModel",
        debug: () => table.options.debugAll ?? table.options.debugRows
      }
    );
  };
}
function rowOrderCompareFn(rowOrder, a, b) {
  let indexA = rowOrder.indexOf(a.id);
  if (indexA === -1) {
    indexA = rowOrder.indexOf(`${a.id}${import_constants.DATA_TABLE_ROW_EXPANDED_ID_SUFFIX}`);
  }
  let indexB = rowOrder.indexOf(b.id);
  if (indexA === -1) {
    indexB = rowOrder.indexOf(`${b.id}${import_constants.DATA_TABLE_ROW_EXPANDED_ID_SUFFIX}`);
  }
  return indexA - indexB;
}
function rowOrderCompareStrFn(rowOrder, a, b) {
  let indexA = rowOrder.indexOf(a);
  if (indexA === -1) {
    indexA = rowOrder.indexOf(`${a}${import_constants.DATA_TABLE_ROW_EXPANDED_ID_SUFFIX}`);
  }
  let indexB = rowOrder.indexOf(b);
  if (indexA === -1) {
    indexB = rowOrder.indexOf(`${b}${import_constants.DATA_TABLE_ROW_EXPANDED_ID_SUFFIX}`);
  }
  return indexA - indexB;
}
