// @dynatrace/strato-components/tables/DataTable/features/RowOrder/row-order-utils.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var row_order_utils_exports = {};
__export(row_order_utils_exports, {
  DATA_TABLE_ROW_DND_COLLISION_OFFSET: () => DATA_TABLE_ROW_DND_COLLISION_OFFSET,
  closestTargetWithinY: () => closestTargetWithinY,
  createRestrictToTableRowAreaModifier: () => createRestrictToTableRowAreaModifier,
  createRowOrderCollisionDetection: () => createRowOrderCollisionDetection,
  getAllNonTargetRowsInTableBody: () => getAllNonTargetRowsInTableBody,
  getAllRowsInTableBody: () => getAllRowsInTableBody,
  getDragHandleForRow: () => getDragHandleForRow,
  getDragIndexes: () => getDragIndexes,
  getRowWithId: () => getRowWithId,
  resetAllStylesForRows: () => resetAllStylesForRows,
  restrictToVerticalBounds: () => restrictToVerticalBounds,
  transformRowOnDragOver: () => transformRowOnDragOver
});
module.exports = __toCommonJS(row_order_utils_exports);
var import_animations = __toESM(require("@dynatrace/strato-design-tokens/animations"));
var import_RowOrder_css = require("./RowOrder.sty.js");
var import_data_table_border_width = require("../../../utils/data-table-border-width.js");
var import_constants = require("../../constants.js");
let lastAffectedRows = /* @__PURE__ */ new Set();
function transformRowOnDragOver(table, event, rowOrder, rowSeparationNone, activeRow) {
  let rowHeight = event.active?.rect.current.translated?.height;
  if (activeRow === void 0 || rowOrder === void 0 || event.collisions === null || event.active === null || event.over === null || rowHeight === void 0) {
    return;
  }
  styleNonTargetRows(table, activeRow.parentId);
  const newAffectedRows = /* @__PURE__ */ new Set();
  const { fromIndex, toIndex } = getDragIndexes(
    rowOrder,
    event.active.id,
    // for us the IDs are always strings
    event.over.id
    // for us the IDs are always strings
  );
  const activeRowIsLastRow = fromIndex === rowOrder.length - 1;
  if (!rowSeparationNone && !activeRowIsLastRow && !activeRow.getIsDetailsOpen()) {
    rowHeight += import_constants.DATA_TABLE_ROW_BORDER;
  }
  if (fromIndex < toIndex) {
    transformRows(
      table,
      rowOrder.slice(fromIndex + 1, toIndex + 1),
      -rowHeight,
      newAffectedRows
    );
  } else if (fromIndex > toIndex) {
    transformRows(
      table,
      rowOrder.slice(toIndex, fromIndex),
      rowHeight,
      newAffectedRows
    );
  }
  transformRows(table, lastAffectedRows, 0);
  lastAffectedRows = newAffectedRows;
}
function transformRows(table, rowIds, amountToTransform, newAffectedRows) {
  for (const rowId of rowIds) {
    const row = getRowWithId(table, rowId);
    if (row.length <= 0) {
      continue;
    }
    translateRowY(row, amountToTransform);
    if (newAffectedRows) {
      lastAffectedRows.delete(rowId);
      newAffectedRows.add(rowId);
    }
  }
}
function styleNonTargetRows(table, activeParentId) {
  if (activeParentId === void 0) {
    return;
  }
  const rows = getAllNonTargetRowsInTableBody(table, activeParentId);
  for (const row of rows) {
    row.classList.add(import_RowOrder_css.disabledNonTargetRow);
  }
}
function translateRowY(row, value) {
  const transform = `translateY(${value}px)`;
  const transition = `transform ${import_animations.default.SlideIn.Duration.Base} ${import_animations.default.Move.Easing}`;
  for (const rowElement of row) {
    rowElement.style.transform = transform;
    rowElement.style.transition = transition;
  }
}
function resetAllStylesForRows(rows) {
  if (rows.length === 0) {
    return;
  }
  for (const row of rows) {
    row.style.removeProperty("transform");
    row.style.removeProperty("transition");
    row.classList.remove(import_RowOrder_css.disabledNonTargetRow);
  }
}
function getRowWithId(table, id) {
  if (id === void 0) {
    return [];
  }
  const rowElements = table.querySelectorAll(
    `[data-rowid="${CSS.escape(String(id))}"], [data-rowid="${CSS.escape(
      String(id)
    )}${import_constants.DATA_TABLE_ROW_EXPANDED_ID_SUFFIX}"], [data-rowid="${CSS.escape(
      String(id)
    )}${import_constants.DATA_TABLE_PINNING_SHADOW_CELL_ROW_ID_SUFFIX}"]`
  );
  return Array.from(rowElements);
}
function getAllNonTargetRowsInTableBody(table, parentIdToExclude) {
  const nonTargetRows = table.querySelectorAll(
    `[data-rowgroup="body"] [role="row"]:not([data-rowid^="${CSS.escape(
      String(parentIdToExclude)
    )}"]`
  );
  return Array.from(nonTargetRows);
}
function getAllRowsInTableBody(table) {
  const allRowsInTableBody = table.querySelectorAll(
    '[role="row"][data-rowid], [role="presentation"][data-rowid]'
  );
  return Array.from(allRowsInTableBody);
}
function getDragHandleForRow(table, rowNumber) {
  return table.querySelector(
    `[role="row"][data-row="${rowNumber}"] [role="cell"][data-column="0"] button`
  );
}
function getDragIndexes(rowOrder, activeId, overId) {
  const fromIndex = rowOrder.findIndex((id) => id === activeId);
  const toIndex = rowOrder.findIndex((id) => id === overId);
  return { fromIndex, toIndex };
}
function isPointWithinRectY(pointY, rect) {
  return rect.top <= pointY && pointY <= rect.bottom;
}
const DATA_TABLE_ROW_DND_COLLISION_OFFSET = import_constants.DATA_TABLE_DEFAULT_MIN_ROW_HEIGHT / 2 - 1;
const closestTargetWithinY = ({
  droppableContainers,
  droppableRects,
  collisionRect,
  active
}) => {
  const initialRect = active.rect.current.initial;
  if (initialRect === null) {
    return [];
  }
  const collisions = [];
  const isDraggingUp = collisionRect.top < initialRect.top;
  const collisionRectY = isDraggingUp ? collisionRect.top + DATA_TABLE_ROW_DND_COLLISION_OFFSET : collisionRect.bottom - DATA_TABLE_ROW_DND_COLLISION_OFFSET;
  for (const droppableContainer of droppableContainers) {
    const { id, node } = droppableContainer;
    const rect = droppableRects.get(id);
    if (rect === void 0) {
      continue;
    }
    const heightAdaptedRect = { ...rect };
    const rowDetailsElement = node.current?.parentElement?.querySelector(
      `[data-rowid="${CSS.escape(
        String(id)
      )}${import_constants.DATA_TABLE_ROW_EXPANDED_ID_SUFFIX}"]`
    );
    if (rowDetailsElement) {
      const detailsHeight = rowDetailsElement.getBoundingClientRect().height;
      heightAdaptedRect.height += detailsHeight;
      heightAdaptedRect.bottom += detailsHeight;
    }
    if (isPointWithinRectY(collisionRectY, heightAdaptedRect)) {
      return [
        {
          id,
          data: { droppableContainer }
        }
      ];
    }
    const droppableContainerY = isDraggingUp ? rect.top : rect.bottom;
    const distance = Math.abs(collisionRectY - droppableContainerY);
    collisions.push({ id, data: { droppableContainer, value: distance } });
  }
  return collisions.toSorted((a, b) => a.data.value - b.data.value);
};
function findLastDisabledRow(rows) {
  let lastDisabledRowIndex = void 0;
  for (const { rowIndex, row } of rows) {
    if (row.getCanDrag()) {
      break;
    }
    lastDisabledRowIndex = rowIndex;
  }
  return lastDisabledRowIndex;
}
const createRowOrderCollisionDetection = (draggingRowInfo) => {
  return ({ active, droppableContainers, ...rest }) => {
    if (draggingRowInfo === void 0) {
      return [];
    }
    const activeRow = draggingRowInfo.row;
    const activeRowIndex = draggingRowInfo.index;
    const activeRowSubRowDepth = activeRow.getSubRowDepth();
    const lastExpandedSubRowsOnLevel = /* @__PURE__ */ new Set();
    const validTargets = droppableContainers.filter((target) => {
      const targetRow = target.data.current?.row;
      const targetRowIndex = target.data.current?.rowIndex;
      const targetRowSubRowDepth = targetRow?.getSubRowDepth();
      const targetRowDisabled = !targetRow?.getCanDrag();
      const targetRowLockedDisabled = targetRow?.disabledRowDragAndDropLockPosition;
      const isSameDepthAndParent = targetRowSubRowDepth === activeRowSubRowDepth && activeRow.parentId === targetRow.parentId;
      if (activeRowIndex === targetRowIndex) {
        return true;
      }
      if (targetRowDisabled && targetRowLockedDisabled !== false) {
        const rowsByRowIndex = droppableContainers.filter(
          (container) => container.data.current?.row.parentId === void 0
        ).map((container) => ({
          rowIndex: container.data.current?.rowIndex,
          row: container.data.current?.row
        })).toSorted((data1, data2) => data1?.rowIndex - data2?.rowIndex);
        const lastDisabledRowIndexFromStart = targetRowLockedDisabled !== "end" ? findLastDisabledRow(rowsByRowIndex) : void 0;
        const firstDisabledRowIndexFromEnd = targetRowLockedDisabled !== "start" ? findLastDisabledRow(rowsByRowIndex.toReversed()) : void 0;
        if (targetRow.parentId === void 0 && (lastDisabledRowIndexFromStart !== void 0 && targetRowIndex <= lastDisabledRowIndexFromStart || firstDisabledRowIndexFromEnd !== void 0 && targetRowIndex >= firstDisabledRowIndexFromEnd)) {
          return false;
        }
      }
      if (isSameDepthAndParent) {
        const hasOpenSubRows = targetRow.getCanExpand() && targetRow.getIsExpanded();
        if (activeRowIndex < targetRowIndex && hasOpenSubRows) {
          const lastExpandedRow = targetRow.getLastExpandedSubRow();
          if (lastExpandedRow) {
            lastExpandedSubRowsOnLevel.add(lastExpandedRow.id);
          }
          return false;
        }
        return true;
      }
      if (targetRowSubRowDepth > activeRowSubRowDepth) {
        return lastExpandedSubRowsOnLevel.has(targetRow.id);
      }
      return false;
    });
    return closestTargetWithinY({
      active,
      droppableContainers: validTargets,
      ...rest
    });
  };
};
function restrictToVerticalBounds(transform, rect, verticalBounds) {
  const value = {
    ...transform
  };
  if (rect.top >= verticalBounds.top && rect.top + transform.y <= verticalBounds.top) {
    value.y = verticalBounds.top - rect.top;
  } else if (rect.bottom <= verticalBounds.bottom && rect.bottom + transform.y >= verticalBounds.bottom) {
    value.y = verticalBounds.bottom - rect.bottom;
  } else if (rect.top < verticalBounds.top) {
    const bottomTransformYLimit = verticalBounds.bottom - rect.bottom;
    const transformY = transform.y > 0 ? verticalBounds.top - rect.top + transform.y : verticalBounds.top - rect.top;
    value.y = transformY > bottomTransformYLimit ? bottomTransformYLimit : transformY;
  } else if (rect.bottom > verticalBounds.bottom) {
    const topTransformYLimit = verticalBounds.top - rect.top;
    const transformY = transform.y < 0 ? verticalBounds.bottom - rect.bottom + transform.y : verticalBounds.bottom - rect.bottom;
    value.y = transformY > topTransformYLimit ? transformY : topTransformYLimit;
  }
  return value;
}
function createRestrictToTableRowAreaModifier(tableHeaderHeight, containedVariant, virtualizationContainer) {
  return ({ transform, draggingNodeRect }) => {
    if (!draggingNodeRect || virtualizationContainer === void 0) {
      return transform;
    }
    transform.x = 0;
    const parentRect = virtualizationContainer.getBoundingClientRect();
    const verticalBounds = {
      top: containedVariant ? parentRect?.top + tableHeaderHeight + (0, import_data_table_border_width.getDataTableBorderWidth)() : parentRect?.top + tableHeaderHeight,
      bottom: containedVariant ? parentRect?.bottom - (0, import_data_table_border_width.getDataTableBorderWidth)() : parentRect?.bottom
    };
    return restrictToVerticalBounds(
      transform,
      draggingNodeRect,
      verticalBounds
    );
  };
}
