// @dynatrace/strato-components/tables/DataTable/features/RowOrder/RowOrder.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var RowOrder_exports = {};
__export(RowOrder_exports, {
  DataTableRowOrder: () => DataTableRowOrder,
  ROW_DRAG_COLUMN: () => ROW_DRAG_COLUMN,
  RowOrderingDragAndDropContext: () => RowOrderingDragAndDropContext,
  useRowOrder: () => useRowOrder
});
module.exports = __toCommonJS(RowOrder_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_core = require("@dnd-kit/core");
var import_sortable = require("@dnd-kit/sortable");
var import_react_table = require("@tanstack/react-table");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_dom = require("react-dom");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_row_order_model = require("./row-order-model.js");
var import_row_order_utils = require("./row-order-utils.js");
var styles = __toESM(require("./RowOrder.sty.js"));
var import_Button = require("../../../../buttons/button/Button.js");
var import_useControllableState = require("../../../../core/hooks/useControllableState.js");
var import_useModality = require("../../../../core/hooks/useModality.js");
var import_focusRingStatic_css = require("../../../../core/styles/focusRingStatic.sty.js");
var import_get_inert_value = require("../../../../core/utils/get-inert-value.js");
var import_logging = require("../../../../core/utils/logging.js");
var import_Grid = require("../../../../layouts/grid/Grid.js");
var import_DataTableDefaultCell = require("../../components/DataTableDefaultCell.js");
var import_DataTableRow = require("../../components/DataTableRow.js");
var import_constants = require("../../constants.js");
var import_DataTableContext = require("../../contexts/DataTableContext.js");
var import_utils = require("../../hooks/useTable/utils.js");
var import_messages = require("../../messages.js");
var import_RowDetailsElement = require("../RowDetails/RowDetailsElement.js");
var thresholdStyles = __toESM(require("../Thresholds/Thresholds.sty.js"));
const SCROLL_JUMP_THRESHOLD = import_constants.DATA_TABLE_ROW_HEIGHT_MAP.default * 10;
function isRowOrderConfigObject(rowOrdering) {
  return typeof rowOrdering === "object";
}
function useRowOrder(props, options) {
  const { rowOrdering } = props;
  const disableRowDragAndDropFn = typeof rowOrdering === "object" ? rowOrdering.disableRowDragAndDrop : void 0;
  const enableRowDragAndDropCallback = (0, import_react.useCallback)(
    (row) => {
      if (!disableRowDragAndDropFn) {
        return true;
      }
      return !disableRowDragAndDropFn(row.original);
    },
    [disableRowDragAndDropFn]
  );
  const [rowOrder, onRowOrderChange] = (0, import_useControllableState.useControllableState)({
    value: props.rowOrder,
    defaultValue: props.defaultRowOrder,
    onChange: props.onRowOrderChange
  });
  options.enableRowOrderDragAndDrop = false;
  if (rowOrdering === false || rowOrdering === void 0) {
    options.enableRowOrdering = false;
    return;
  }
  options.enableRowOrdering = true;
  options.getRowOrderRowModel = (0, import_row_order_model.getRowOrderRowModel)();
  options.onRowOrderChange = onRowOrderChange;
  options.state = { ...options.state, rowOrder };
  if (isRowOrderConfigObject(rowOrdering)) {
    if (rowOrdering.enableDragAndDrop === true && rowOrdering.disableRowDragAndDrop !== void 0) {
      options.lockDisabledRows = rowOrdering.lockDisabledRows;
      options.enableRowOrderDragAndDrop = enableRowDragAndDropCallback;
      options.autoResetPageIndex = false;
    } else {
      options.enableRowOrderDragAndDrop = rowOrdering.enableDragAndDrop === true;
    }
  }
}
const RowDragHandleCell = ({
  rowId,
  disabled,
  isDraggingActive
}) => {
  const intl = (0, import_react_intl.useIntl)();
  const modality = (0, import_useModality.useModality)();
  const { attributes, listeners } = (0, import_sortable.useSortable)({
    id: rowId
  });
  const isKeyboardDragging = isDraggingActive && modality === "keyboard";
  const rowDragHandleLabel = intl.formatMessage(
    import_messages.dataTableMessages.rowDragAndDropLabel,
    { rowId }
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_DataTableDefaultCell.DataTableDefaultCell, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Button.Button,
    {
      size: "condensed",
      variant: "default",
      ...attributes,
      ...!disabled ? listeners : void 0,
      "aria-disabled": disabled,
      "aria-label": rowDragHandleLabel,
      className: (0, import_clsx.default)(
        styles.dndHandle({
          disabled
        }),
        // Apply the focus ring unconditionally (when: 'always')
        // to the dragged row's handle during keyboard dragging.
        isKeyboardDragging && (0, import_focusRingStatic_css.focusRingStatic)({
          minimal: false,
          color: "neutral",
          when: "always"
        })
      ),
      "data-dt-name": "RowOrderDragAndDrop",
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.DragAllDirectionIcon, {}) })
    }
  ) });
};
const ROW_DRAG_COLUMN = {
  id: import_constants.DATA_TABLE_ROW_DRAG_COLUMN_ID,
  alignment: "center",
  accessorFn: () => "",
  header: () => {
    return "";
  },
  width: "content",
  minWidth: import_constants.DATA_TABLE_BUILTIN_COLUMN_MIN_WIDTH,
  enableSorting: false,
  enableResizing: false,
  enableUserActions: false,
  enableHiding: false,
  meta: {
    builtin: true,
    hasCustomCellRenderer: true,
    hasCustomHeaderRenderer: true,
    builtinPlacement: "prefix",
    cellDeps: (cell) => {
      return [cell.row.getCanDrag()];
    }
  },
  cell: ({ row, table }) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    RowDragHandleCell,
    {
      rowId: row.id,
      disabled: !row.getCanDrag(),
      isDraggingActive: table.getState().draggingRowInfo?.id === row.id
    }
  )
};
const DataTableRowOrder = {
  createTable(table) {
    table.getPreRowOrderRowModel = () => table.getExpandedRowModel();
    table.setDraggingRowInfo = (draggingRowInfo) => {
      table.setState((state) => ({
        ...state,
        draggingRowInfo
      }));
    };
    table.getRowOrderRowModel = () => {
      if (!table._getRowOrderRowModel && table.options.getRowOrderRowModel) {
        table._getRowOrderRowModel = table.options.getRowOrderRowModel(table);
      }
      if (!table._getRowOrderRowModel) {
        return table.getPreRowOrderRowModel();
      }
      return table._getRowOrderRowModel();
    };
    table.setRowOrder = (updater) => {
      const newValue = (0, import_react_table.functionalUpdate)(
        updater,
        table.getState().rowOrder ?? []
      );
      table._queue(() => {
        table.options.onRowOrderChange?.(newValue);
      });
    };
    (0, import_utils.assignRowMethods)(table, {
      getCanDrag() {
        const { enableRowOrdering, enableRowOrderDragAndDrop } = table.options;
        const sortingState = table.getState().sorting;
        if (enableRowOrdering !== true || // always disable the drag and drop handle if the rows are sorted
        sortingState !== void 0 && sortingState.length > 0) {
          return false;
        }
        if (enableRowOrderDragAndDrop !== void 0 && typeof enableRowOrderDragAndDrop === "function") {
          return enableRowOrderDragAndDrop(this);
        }
        return true;
      },
      disabledRowDragAndDropLockPosition: table.options.lockDisabledRows ?? false
    });
  }
};
function RowOrderingDragAndDropContext({
  children,
  rows,
  virtualColumns,
  rowSeparation
}) {
  const table = (0, import_DataTableContext.useDataTable)();
  const modality = (0, import_useModality.useModality)();
  const tableHeaderHeight = table.getTotalHeaderHeight();
  const sensors = (0, import_core.useSensors)(
    (0, import_core.useSensor)(import_core.MouseSensor, {}),
    (0, import_core.useSensor)(import_core.TouchSensor, {}),
    (0, import_core.useSensor)(import_core.KeyboardSensor, {})
  );
  const [draggingElement, setDraggingElement] = (0, import_react.useState)(
    void 0
  );
  const containedVariant = table.getVisualVariantOptions().contained;
  const virtualizationContainer = table.getRowVirtualizationContainer();
  const draggingRowInfo = table.getState().draggingRowInfo;
  const tableGridElement = virtualizationContainer?.querySelector("[role=grid]");
  const overlayContainer = (0, import_react.useRef)(null);
  const scrollHandlerRef = (0, import_react.useRef)(null);
  const rowOrderCollisionDetection = (0, import_react.useMemo)(
    () => (0, import_row_order_utils.createRowOrderCollisionDetection)(draggingRowInfo),
    [draggingRowInfo]
  );
  const restrictToTableRowArea = (0, import_react.useMemo)(
    () => (0, import_row_order_utils.createRestrictToTableRowAreaModifier)(
      tableHeaderHeight,
      containedVariant,
      virtualizationContainer
    ),
    [tableHeaderHeight, virtualizationContainer, containedVariant]
  );
  const rowOrderWhileDragging = (0, import_react.useRef)([]);
  if (!tableGridElement || !virtualizationContainer) {
    return;
  }
  const overlayWidth = tableGridElement.parentElement?.clientWidth ?? 0;
  const originalRowOrder = table.getState().rowOrder ?? rows.map((row) => row.id);
  const resetDragging = () => {
    if (virtualizationContainer && scrollHandlerRef.current) {
      virtualizationContainer.removeEventListener(
        "scroll",
        scrollHandlerRef.current
      );
      scrollHandlerRef.current = null;
    }
    table.setDraggingRowInfo(void 0);
    setDraggingElement(void 0);
    (0, import_row_order_utils.resetAllStylesForRows)((0, import_row_order_utils.getAllRowsInTableBody)(tableGridElement));
  };
  const handleOnDragStart = ({ active }) => {
    if (active.data.current) {
      const element = createDraggingElement(
        overlayContainer,
        active,
        rows,
        virtualColumns,
        {
          width: overlayWidth,
          gridTemplateColumns: table.getGridTemplateColumns(),
          rowSeparation,
          hasThresholds: table.options.hasThresholds,
          rowThresholds: table.options.rowThresholds,
          rowDetailsTemplate: table.options.rowDetailsTemplate
        }
      );
      setDraggingElement(element);
      requestAnimationFrame(() => {
        if (overlayContainer.current?.parentElement) {
          overlayContainer.current.parentElement.scrollLeft = tableGridElement?.parentElement?.scrollLeft ?? 0;
        }
      });
      table.setDraggingRowInfo({
        id: active.id,
        index: active.data.current.rowIndex,
        row: active.data.current.row
      });
      rowOrderWhileDragging.current = originalRowOrder ? [...originalRowOrder] : [];
      let lastScrollOffset = virtualizationContainer?.scrollTop ?? 0;
      const handleScroll = () => {
        if (!virtualizationContainer) {
          return;
        }
        const currentOffset = virtualizationContainer.scrollTop;
        const scrollDelta = Math.abs(currentOffset - lastScrollOffset);
        if (scrollDelta >= SCROLL_JUMP_THRESHOLD) {
          virtualizationContainer.scrollTop = lastScrollOffset;
          return;
        }
        lastScrollOffset = currentOffset;
      };
      scrollHandlerRef.current = handleScroll;
      virtualizationContainer?.addEventListener("scroll", handleScroll);
    }
  };
  const handleOnDragEnd = ({
    active,
    over
  }) => {
    if (over === void 0 || over === null) {
      (0, import_logging.devLog)("[Table] On drag end: over info is undefined or null", over);
      resetDragging();
      return;
    }
    if (draggingRowInfo === void 0) {
      (0, import_logging.devLog)(
        "[Table] On drag end: dragging row info is undefined",
        draggingRowInfo
      );
      resetDragging();
      return;
    }
    if (over.id === active.id) {
      resetDragging();
      return;
    }
    const { fromIndex, toIndex } = (0, import_row_order_utils.getDragIndexes)(
      originalRowOrder,
      active.id,
      // for us the IDs are always strings
      over.id
      // for us the IDs are always strings
    );
    const newRowOrder = [...originalRowOrder];
    const activeRowSubRowsCount = draggingRowInfo.row.getLeafRows().length;
    const overRowSubRowsCount = (over.data.current?.row).getLeafRows().length;
    const rowsToMove = newRowOrder.splice(fromIndex, 1 + activeRowSubRowsCount);
    newRowOrder.splice(
      fromIndex < toIndex ? toIndex + overRowSubRowsCount - activeRowSubRowsCount : toIndex,
      0,
      ...rowsToMove
    );
    table.setRowOrder(newRowOrder);
    resetDragging();
    virtualizationContainer?.scrollBy({ top: 1, behavior: "smooth" });
    virtualizationContainer?.scrollBy({ top: -1, behavior: "smooth" });
    requestAnimationFrame(() => {
      const dragHandle = (0, import_row_order_utils.getDragHandleForRow)(tableGridElement, toIndex);
      dragHandle?.focus(
        modality !== "keyboard" ? { focusVisible: false } : void 0
      );
    });
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_core.DndContext,
    {
      sensors,
      collisionDetection: rowOrderCollisionDetection,
      modifiers: [restrictToTableRowArea],
      onDragStart: handleOnDragStart,
      onDragEnd: handleOnDragEnd,
      onDragOver: (event) => (0, import_row_order_utils.transformRowOnDragOver)(
        tableGridElement,
        event,
        originalRowOrder,
        table.getVisualVariantOptions().rowSeparation === "none",
        draggingRowInfo?.row
      ),
      onDragCancel: () => resetDragging(),
      children: [
        children,
        // We need to portal the drag and drop overlay to escape CSS transforms on ancestor elements.
        // This ensures transforms on parent containers don't affect the position of the drag overlay.
        (0, import_react_dom.createPortal)(
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_core.DragOverlay,
            {
              dropAnimation: null,
              className: styles.rowDndOverlay,
              style: { width: overlayWidth, height: "max-content" },
              children: draggingElement
            }
          ),
          document.body
        )
      ]
    }
  );
}
function createDraggingElement(overlayContainer, active, rows, virtualColumns, rowConfig) {
  const activeId = active.id;
  const draggingOriginalRow = active.data.current?.row;
  const draggingVirtualRow = active.data.current?.virtualRow;
  const {
    gridTemplateColumns,
    rowSeparation,
    hasThresholds,
    rowThresholds,
    rowDetailsTemplate
  } = rowConfig;
  if (draggingOriginalRow === void 0 || draggingVirtualRow === void 0) {
    (0, import_logging.devLog)(
      `[DataTable] Dragging row or virtual row with row id ${activeId} is undefined.`
    );
    return;
  }
  const isRowDetailsOpen = draggingOriginalRow.getIsDetailsOpen();
  const draggingVirtualDetailsRow = {
    ...draggingVirtualRow,
    index: draggingVirtualRow.index + 1
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_Grid.Grid,
    {
      ref: overlayContainer,
      gridTemplateColumns,
      padding: 0,
      minWidth: "max-content",
      gap: 0,
      ...{ inert: (0, import_get_inert_value.getProperInertValue)(true) },
      className: (0, import_clsx.default)({
        [thresholdStyles.hasPillThresholds]: hasThresholds && rowThresholds?.some((threshold) => threshold.type === "pill")
      }),
      style: {
        position: "relative"
      },
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_DataTableRow.DataTableRow,
          {
            row: draggingOriginalRow,
            virtualRow: draggingVirtualRow,
            virtualItemIndex: 0,
            rowCount: rows.length,
            virtualColumns,
            isDraggedRow: false,
            isInDragOverlay: true,
            rowParity: "even"
          }
        ),
        isRowDetailsOpen ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_RowDetailsElement.RowDetailsElement,
          {
            originalRow: draggingOriginalRow,
            virtualRow: draggingVirtualDetailsRow,
            virtualItemIndex: 1,
            rowSeparation,
            rowParity: "even",
            children: rowDetailsTemplate?.({
              row: draggingOriginalRow.original
            })
          }
        ) : null
      ]
    }
  );
}
