// @dynatrace/strato-components/esm/charts/core/components/sparkline-display/components/ContextValueIndicator.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import { TextEllipsis } from "../../../../../typography/text-ellipsis/TextEllipsis.js";
import { TestIds } from "../../annotations/constants.js";
import { DEFAULT_STROKE_WIDTH } from "../../grid/GridLine.js";
import { charsFontWidth12pixelsSize500Map } from "../../text-ellipsis/chars-font-width-map.js";
import { calculateWidth } from "../../text-ellipsis/truncate-string.js";
import {
  CONTEXT_VALUE_STROKE_COLOR,
  CONTEXT_VALUE_STROKE_DASHARRAY,
  CONTEXT_VALUE_STROKE_LINECAP,
  CONTEXT_VALUE_TEXT_OFFSET_MAX,
  CONTEXT_VALUE_TEXT_OFFSET_MIN_LARGE,
  CONTEXT_VALUE_TEXT_OFFSET_MIN_SMALL,
  CONTEXT_VALUES_Y_OFFSET,
  NEGATIVE_SIGN
} from "../constants.js";
import { contextValueText, negativeSign } from "../Sparkline.sty.js";
import {
  SparklineContainerHeightFontSize
} from "../types/sparkline-container-height.js";
import { getContextValueXOffset } from "../utils/get-context-value-x-offset.js";
function getTextOffset(isMax, fontSize) {
  if (isMax) {
    return CONTEXT_VALUE_TEXT_OFFSET_MAX;
  }
  return fontSize === SparklineContainerHeightFontSize.SMALLER_THAN_32 ? CONTEXT_VALUE_TEXT_OFFSET_MIN_SMALL : CONTEXT_VALUE_TEXT_OFFSET_MIN_LARGE;
}
function computeIndicatorPositioning(props) {
  const {
    formattedValue,
    yPosition,
    dottedLineWidth,
    containerHeight,
    fontSize,
    labelsAlignment,
    chartLeftOffset,
    labelsFixedWidth,
    labelYPosition,
    isMax = false
  } = props;
  const isNegative = formattedValue.startsWith(NEGATIVE_SIGN);
  const displayValue = isNegative ? formattedValue.slice(1) : formattedValue;
  const isLeft = labelsAlignment !== "right";
  const actualWidth = calculateWidth(
    displayValue,
    charsFontWidth12pixelsSize500Map
  );
  const labelWidth = typeof labelsFixedWidth === "number" && !Number.isNaN(labelsFixedWidth) ? labelsFixedWidth : actualWidth;
  const positioningWidth = isLeft ? Math.min(actualWidth, labelWidth) : labelWidth;
  const { signXOffset, textXOffset } = getContextValueXOffset(
    isLeft,
    chartLeftOffset,
    positioningWidth,
    dottedLineWidth,
    containerHeight,
    fontSize === SparklineContainerHeightFontSize.SMALLER_THAN_32
  );
  const textOffset = getTextOffset(isMax, fontSize);
  const labelYDelta = (labelYPosition ?? yPosition) - yPosition;
  return {
    containerTransform: `translateY(${yPosition}px) translateX(${isLeft ? 0 : chartLeftOffset}px)`,
    svgLeft: isLeft ? chartLeftOffset : 0,
    labelYTransform: `translateY(${labelYDelta - textOffset}px)`,
    signXOffset,
    textXOffset,
    actualWidth,
    isNegative,
    displayValue
  };
}
const ContextValueIndicator = (props) => {
  const { dottedLineWidth, containerHeight, fontSize, color } = props;
  const {
    containerTransform,
    svgLeft,
    labelYTransform,
    signXOffset,
    textXOffset,
    actualWidth,
    isNegative,
    displayValue
  } = computeIndicatorPositioning(props);
  return /* @__PURE__ */ jsxs(
    "div",
    {
      className: contextValueText({ fontSize }),
      style: { transform: containerTransform },
      children: [
        /* @__PURE__ */ jsx(
          "svg",
          {
            width: dottedLineWidth,
            style: { position: "absolute", left: svgLeft, top: 0 },
            children: /* @__PURE__ */ jsx(
              "line",
              {
                "data-testid": TestIds.AnnotationIndicator,
                x1: 0,
                x2: dottedLineWidth,
                y1: CONTEXT_VALUES_Y_OFFSET,
                y2: CONTEXT_VALUES_Y_OFFSET,
                stroke: CONTEXT_VALUE_STROKE_COLOR,
                strokeWidth: DEFAULT_STROKE_WIDTH,
                strokeDasharray: CONTEXT_VALUE_STROKE_DASHARRAY,
                strokeLinecap: CONTEXT_VALUE_STROKE_LINECAP
              }
            )
          }
        ),
        /* @__PURE__ */ jsx(
          "div",
          {
            className: clsx(negativeSign({ containerHeight })),
            style: {
              transform: `${labelYTransform} translateX(${signXOffset}px)`
            },
            children: isNegative && NEGATIVE_SIGN
          }
        ),
        /* @__PURE__ */ jsx(
          TextEllipsis,
          {
            truncationMode: "middle",
            style: {
              width: `${actualWidth}px`,
              position: "absolute",
              transform: `${labelYTransform} translateX(${textXOffset}px)`,
              color
            },
            children: displayValue
          }
        )
      ]
    }
  );
};
export {
  ContextValueIndicator
};
