// @dynatrace/strato-components/esm/tables/convert-to-columns/convert-to-columns.js
import {
  units
} from "@dynatrace-sdk/units";
function extractColumnTypes(rangedFieldTypes) {
  const columnTypes = /* @__PURE__ */ new Map();
  for (const { mappings } of rangedFieldTypes) {
    for (const key in mappings) {
      const type = mappings[key]?.type ?? "string";
      if (columnTypes.has(key)) {
        columnTypes.get(key)?.add(type);
      } else {
        columnTypes.set(key, /* @__PURE__ */ new Set([type]));
      }
    }
  }
  return columnTypes;
}
const FieldTypeTypeToColumnType = /* @__PURE__ */ new Map([
  ["timestamp", "datetime"],
  ["long", "long"],
  ["double", "number"],
  ["duration", "number"]
]);
const FieldTypeTypeToFormatter = /* @__PURE__ */ new Map([
  [
    "duration",
    {
      input: units.time.nanosecond,
      cascade: 3
    }
  ]
]);
function isTimeFrameEntry(value) {
  if (value === null || typeof value !== "object") {
    return false;
  }
  const { start, end } = value;
  return (typeof start === "string" || start instanceof Date) && (typeof end === "string" || end instanceof Date);
}
function isTimeGeoPointEntry(value) {
  if (value === null || typeof value !== "object") {
    return false;
  }
  const { latitude, longitude } = value;
  return typeof latitude === "number" && typeof longitude === "number";
}
function getTimeframeColumn(columnKey) {
  return {
    header: columnKey,
    id: columnKey,
    columns: [
      {
        header: "start",
        id: `${columnKey}.start`,
        accessor: (row) => {
          const entry = row?.[columnKey];
          if (isTimeFrameEntry(entry)) {
            const rawDate = entry.start;
            if (rawDate instanceof Date) {
              return rawDate;
            }
            const date = new Date(rawDate);
            return Number.isNaN(date.getTime()) ? "" : date;
          }
          return "";
        },
        columnType: "datetime",
        width: "content"
      },
      {
        header: "end",
        id: `${columnKey}.end`,
        accessor: (row) => {
          const entry = row?.[columnKey];
          if (isTimeFrameEntry(entry)) {
            const rawDate = entry.end;
            if (rawDate instanceof Date) {
              return rawDate;
            }
            const date = new Date(rawDate);
            return Number.isNaN(date.getTime()) ? "" : date;
          }
          return "";
        },
        columnType: "datetime",
        width: "content"
      }
    ]
  };
}
function getGeoPointColumn(columnKey) {
  return {
    header: columnKey,
    id: columnKey,
    columns: [
      {
        header: "longitude",
        id: `${columnKey}.longitude`,
        accessor: (row) => {
          if (isTimeGeoPointEntry(row?.[columnKey])) {
            return row[columnKey].longitude;
          }
          return "";
        },
        columnType: "text",
        width: "content"
      },
      {
        header: "latitude",
        id: `${columnKey}.latitude`,
        accessor: (row) => {
          if (isTimeGeoPointEntry(row?.[columnKey])) {
            return row[columnKey].latitude;
          }
          return "";
        },
        columnType: "text",
        width: "content"
      }
    ]
  };
}
function convertToColumns(rangedFieldTypes) {
  const extractedTypes = extractColumnTypes(rangedFieldTypes);
  const columns = [];
  for (const [columnKey, columnTypes] of extractedTypes.entries()) {
    const columnHasOnlyOneType = columnTypes.size === 1;
    const columnType = columnHasOnlyOneType ? Array.from(columnTypes)[0] : "string";
    if (columnType === "timeframe") {
      columns.push(getTimeframeColumn(columnKey));
    } else if (columnType === "geo_point") {
      columns.push(getGeoPointColumn(columnKey));
    } else {
      columns.push({
        header: columnKey,
        id: columnKey,
        accessor: (row) => row?.[columnKey],
        width: "content",
        columnType: FieldTypeTypeToColumnType.get(columnType) ?? "text",
        formatter: FieldTypeTypeToFormatter.get(columnType) ?? void 0
      });
    }
  }
  return columns;
}
export {
  convertToColumns
};
