// @dynatrace/strato-components/core/index.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var core_exports = {};
__export(core_exports, {
  AppRoot: () => import_AppRoot.AppRoot,
  FocusScope: () => import_FocusScope.FocusScope,
  Intent: () => import_Intent.Intent,
  ModalContext: () => import_ModalContext.ModalContext,
  ModalContextProvider: () => import_ModalContextProvider.ModalContextProvider,
  OverlayContainer: () => import_OverlayContainer.OverlayContainer,
  OverlayProvider: () => import_OverlayProvider.OverlayProvider,
  OverlayStackingContext: () => import_OverlayStacking.OverlayStackingContext,
  OverlayStackingProvider: () => import_OverlayStackingProvider.OverlayStackingProvider,
  ScrollProvider: () => import_ScrollProvider.ScrollProvider,
  SharedInputPropsContext: () => import_SharedInputPropsContext.SharedInputPropsContext,
  TIMEFRAME_EXPRESSION: () => import_timeParser.TIMEFRAME_EXPRESSION,
  _FocusProvider: () => import_FocusProvider.FocusProvider,
  _InteractiveSelector: () => import_interactiveSelector._InteractiveSelector,
  _VisuallyHidden: () => import_VisuallyHidden.VisuallyHidden,
  _accessValue: () => import_access_value.accessValue,
  _attemptFocus: () => import_attempt_focus.attemptFocus,
  _canReceiveDefaultTextFormat: () => import_can_receive_default_text_format.canReceiveDefaultTextFormat,
  _checkCombinedThresholdRules: () => import_threshold_utils._checkCombinedThresholdRules,
  _convertToNumberIfPureNumber: () => import_string_utils.convertToNumberIfPureNumber,
  _downloadBlob: () => import_download.downloadBlob,
  _downloadCSV: () => import_download.downloadCSV,
  _downloadFromUrl: () => import_download.downloadFromUrl,
  _dtVisualRefreshEnabled: () => import_AppRoot_css.dtVisualRefreshEnabled,
  _focusFirstDescendant: () => import_focus_first_descendant.focusFirstDescendant,
  _formatDate: () => import_default_formatters.formatDatetimeDefault,
  _getFirstFocusableChild: () => import_get_first_focusable_child.getFirstFocusableChild,
  _getLastFocusableChild: () => import_get_last_focusable_child.getLastFocusableChild,
  _isBrowser: () => import_isBrowser.isBrowser,
  _isFocusable: () => import_is_focusable.isFocusable,
  _matchesThreshold: () => import_threshold_utils._matchesThreshold,
  _mergeProps: () => import_merge_props.mergeProps,
  _parseBoolean: () => import_parse_boolean.parseBoolean,
  _sanitizeSymbol: () => import_sanitize_symbol._sanitizeSymbol,
  _useAriaLabelingProps: () => import_useAriaLabelingProps.useAriaLabelingProps,
  _useBehavioralTrackingProps: () => import_useBehavioralTrackingProps.useBehavioralTrackingProps,
  _useControllableState: () => import_useControllableState.useControllableState,
  _useFocusContext: () => import_useFocusContext.useFocusContext,
  _useFocusRing: () => import_useFocusRing2.useFocusRing,
  _useFormControlState: () => import_useFormControlState.useFormControlState,
  _useIsInViewport: () => import_useIsInViewport.useIsInViewport,
  _useMeasureChildren: () => import_useMeasureChildren.useMeasureChildren,
  _useRequestAnimationFrame: () => import_useRequestAnimationFrame.useRequestAnimationFrame,
  _useSafeForwardProps: () => import_useSafeForwardProps.useSafeForwardProps,
  _useSharedInputPropsContext: () => import_useSharedInputPropsContext.useSharedInputPropsContext,
  createStore: () => import_store.createStore,
  getThresholdColor: () => import_threshold_utils.getThresholdColor,
  isNumber: () => import_type_guards.isNumber,
  isPotentialDateValue: () => import_type_guards.isPotentialDateValue,
  parseTime: () => import_timeParser.parseTime,
  useBreakpointProps: () => import_useBreakpointProps.useBreakpointProps,
  useContainerBreakpoint: () => import_useContainerBreakpoint.useContainerBreakpoint,
  useContainerSize: () => import_useContainerSize.useContainerSize,
  useCurrentTheme: () => import_useCurrentTheme.useCurrentTheme,
  useFocusRing: () => import_useFocusRing.useFocusRing,
  useId: () => import_useId.useId,
  useMergeRefs: () => import_useMergeRefs.useMergeRefs,
  useOverlayStacking: () => import_useOverlayStacking.useOverlayStacking,
  usePermissions: () => import_usePermissions.usePermissions,
  usePrevious: () => import_usePrevious.usePrevious,
  useRovingFocus: () => import_useRovingFocus.useRovingFocus,
  useSafeLayoutEffect: () => import_useSafeLayoutEffect.useSafeLayoutEffect,
  useStoreSelector: () => import_selector.useStoreSelector
});
module.exports = __toCommonJS(core_exports);
var import_AppRoot = require("./components/app-root/AppRoot.js");
var import_AppRoot_css = require("./components/app-root/AppRoot.sty.js");
var import_FocusScope = require("./components/focus-scope/FocusScope.js");
var import_useCurrentTheme = require("./hooks/useCurrentTheme.js");
var import_useFocusContext = require("./hooks/useFocusContext.js");
var import_useFocusRing = require("./styles/useFocusRing.js");
var import_FocusProvider = require("./providers/FocusProvider.js");
var import_attempt_focus = require("./utils/focus-management/attempt-focus.js");
var import_focus_first_descendant = require("./utils/focus-management/focus-first-descendant.js");
var import_get_first_focusable_child = require("./utils/focus-management/get-first-focusable-child.js");
var import_get_last_focusable_child = require("./utils/focus-management/get-last-focusable-child.js");
var import_is_focusable = require("./utils/focus-management/is-focusable.js");
var import_isBrowser = require("./utils/isBrowser.js");
var import_merge_props = require("./utils/merge-props.js");
var import_useAriaLabelingProps = require("./hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("./hooks/useBehavioralTrackingProps.js");
var import_useId = require("./hooks/useId.js");
var import_useSafeForwardProps = require("./hooks/useSafeForwardProps.js");
var import_can_receive_default_text_format = require("./utils/can-receive-default-text-format.js");
var import_parse_boolean = require("./utils/parse-boolean.js");
var import_OverlayContainer = require("./components/OverlayContainer.js");
var import_VisuallyHidden = require("./components/visually-hidden/VisuallyHidden.js");
var import_ModalContext = require("./contexts/ModalContext.js");
var import_OverlayStacking = require("./contexts/OverlayStacking.js");
var import_SharedInputPropsContext = require("./contexts/SharedInputPropsContext.js");
var import_useBreakpointProps = require("./hooks/useBreakpointProps.js");
var import_useContainerBreakpoint = require("./hooks/useContainerBreakpoint.js");
var import_useContainerSize = require("./hooks/useContainerSize.js");
var import_useControllableState = require("./hooks/useControllableState.js");
var import_useFocusRing2 = require("./styles/useFocusRing.js");
var import_useFormControlState = require("./hooks/useFormControlState.js");
var import_useIsInViewport = require("./hooks/useIsInViewport.js");
var import_useMeasureChildren = require("./hooks/useMeasureChildren.js");
var import_useMergeRefs = require("./hooks/useMergeRefs.js");
var import_useOverlayStacking = require("./hooks/useOverlayStacking.js");
var import_usePrevious = require("./hooks/usePrevious.js");
var import_useRequestAnimationFrame = require("./hooks/useRequestAnimationFrame.js");
var import_useRovingFocus = require("./hooks/useRovingFocus.js");
var import_useSafeLayoutEffect = require("./hooks/useSafeLayoutEffect.js");
var import_useSharedInputPropsContext = require("./hooks/useSharedInputPropsContext.js");
var import_ModalContextProvider = require("./providers/ModalContextProvider.js");
var import_OverlayProvider = require("./providers/OverlayProvider.js");
var import_OverlayStackingProvider = require("./providers/OverlayStackingProvider.js");
var import_ScrollProvider = require("./providers/ScrollProvider.js");
var import_selector = require("./store/selector.js");
var import_store = require("./store/store.js");
var import_type_guards = require("./types/type-guards.js");
var import_default_formatters = require("./utils/default-formatters.js");
var import_sanitize_symbol = require("./utils/sanitize-symbol.js");
var import_string_utils = require("./utils/string-utils.js");
var import_interactiveSelector = require("./utils/styles/interactiveSelector.js");
var import_threshold_utils = require("./utils/threshold-utils.js");
var import_timeParser = require("./utils/timeParser.js");
var import_download = require("./utils/download.js");
var import_access_value = require("./utils/accessors/access-value.js");
var import_Intent = require("./slots/Intent/Intent.js");
var import_usePermissions = require("./hooks/usePermissions.js");
