// @dynatrace/strato-components/navigation/tabs/RenderTab.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var RenderTab_exports = {};
__export(RenderTab_exports, {
  RenderTab: () => RenderTab
});
module.exports = __toCommonJS(RenderTab_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_RenderTab_css = require("./RenderTab.sty.js");
var import_useFocusRing = require("../../core/styles/useFocusRing.js");
var import_Text = require("../../typography/text/Text.js");
const RenderTab = (props) => {
  const {
    isDisabled: propsDisabled = false,
    isSelected = false,
    tabItem,
    ariaControls,
    onActivate,
    onKeyDown
  } = props;
  const ref = (0, import_react.useRef)(null);
  const { key, props: tabProps } = tabItem;
  const {
    title,
    disabled,
    children,
    prefixIcon,
    id = key ?? void 0,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    keepMounted,
    ...restProps
  } = tabProps;
  const isDisabled = propsDisabled || (disabled ?? false);
  const { focusProps, focusClassName } = (0, import_useFocusRing.useFocusRing)({
    color: !isDisabled && isSelected ? "primary" : "neutral",
    disabled: isDisabled
  });
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: (0, import_clsx.default)({ [import_RenderTab_css.disabledTabWrapperCSS]: isDisabled }), children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "button",
    {
      "data-testid": dataTestId,
      ...restProps,
      ref,
      type: "button",
      "data-key": key,
      role: "tab",
      id,
      disabled: isDisabled,
      "aria-selected": isSelected,
      "aria-controls": ariaControls,
      ...!isDisabled && !propsDisabled && {
        onClick: onActivate,
        onKeyDown,
        onBlur: focusProps.onBlur,
        onFocus: focusProps.onFocus
      },
      className: (0, import_clsx.default)(
        (0, import_RenderTab_css.tabButtonCSS)({
          disabled: isDisabled,
          selected: isSelected
        }),
        focusClassName,
        consumerClassName
      ),
      style: consumerStyle,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { className: import_RenderTab_css.tabContentWrapperCSS, children: [
        prefixIcon !== void 0 && prefixIcon,
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { className: "text-content", as: "div", textStyle: "base", children: title })
      ] })
    }
  ) });
};
