// @dynatrace/strato-components/navigation/tabs/TabPanel.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TabPanel_exports = {};
__export(TabPanel_exports, {
  TabPanel: () => TabPanel
});
module.exports = __toCommonJS(TabPanel_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_TabPanel_css = require("./TabPanel.sty.js");
var import_useFocusRing = require("../../core/styles/useFocusRing.js");
var import_get_first_focusable_child = require("../../core/utils/focus-management/get-first-focusable-child.js");
var import_merge_props = require("../../core/utils/merge-props.js");
const TabPanel = (props) => {
  const {
    children,
    disabled,
    panelOverflow,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    ...restProps
  } = props;
  const ref = (0, import_react.useRef)(null);
  const { focusProps, focusClassName } = (0, import_useFocusRing.useFocusRing)({ disabled });
  const [hasFocusableChildren, setHasFocusableChildren] = (0, import_react.useState)(false);
  (0, import_react.useLayoutEffect)(() => {
    if (ref.current) {
      setHasFocusableChildren((0, import_get_first_focusable_child.getFirstFocusableChild)(ref.current) !== null);
    }
  }, [children]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      "data-testid": dataTestId,
      className: (0, import_clsx.default)(
        consumerClassName,
        focusClassName,
        import_TabPanel_css.tabPanelCSS,
        import_TabPanel_css.tabPanelOverflowBehaviorCSS[panelOverflow]
      ),
      style: consumerStyle,
      ref,
      ...(0, import_merge_props.mergeProps)(focusProps, restProps),
      tabIndex: hasFocusableChildren ? -1 : 0,
      role: "tabpanel",
      children: !disabled && children
    }
  );
};
TabPanel["displayName"] = "TabPanel";
