// @dynatrace/strato-components/navigation/tabs/Tabs.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Tabs_exports = {};
__export(Tabs_exports, {
  Tabs: () => Tabs
});
module.exports = __toCommonJS(Tabs_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_react_is = require("react-is");
var import_use_resize_observer = __toESM(require("use-resize-observer"));
var import_strato_icons = require("@dynatrace/strato-icons");
var import_RenderTab = require("./RenderTab.js");
var import_Tab = require("./Tab.js");
var import_TabPanel = require("./TabPanel.js");
var import_Tabs_css = require("./Tabs.sty.js");
var import_Button = require("../../buttons/button/Button.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useControllableState = require("../../core/hooks/useControllableState.js");
var import_useId = require("../../core/hooks/useId.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_useRovingFocus = require("../../core/hooks/useRovingFocus.js");
var import_is_node_environment = require("../../core/utils/is-node-environment.js");
var import_is_slot = require("../../core/utils/is-slot.js");
const COMPONENT_NAME = "Tab";
const Tabs = /* @__PURE__ */ (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const {
      disabled: disabledTabsProp = false,
      selectedIndex: selectedIndexProp,
      defaultIndex: defaultIndexProp,
      onChange: onChangeProp,
      children,
      panelOverflow = "hidden",
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...remainingProps
    } = props;
    const intl = (0, import_react_intl.useIntl)();
    const generatedId = (0, import_useId.useId)();
    const tabsUniqueId = props.id ?? generatedId;
    const tabsListId = (0, import_useId.useId)();
    const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
    const childrenArray = (0, import_react.useMemo)(() => {
      return import_react.Children.map(children, (child, i) => {
        if ((0, import_is_slot.isSlot)(child, import_Tab.Tab)) {
          return (0, import_react.cloneElement)(child, {
            key: `${tabsUniqueId}-tab-${i}`
          });
        } else if ((0, import_react_is.isElement)(child) && (0, import_is_node_environment.isDev)()) {
          console.warn("Only Tab is a valid child of the Tabs component.");
        }
      }) || [];
    }, [children, tabsUniqueId]);
    const tabsInfo = (0, import_react.useMemo)(() => {
      return childrenArray.reduce(
        (aggr, item) => {
          if (item.props.id && item.key) {
            aggr.consumerIds[item.key] = item.props.id;
          }
          if (item.props.disabled && item.key !== null) {
            aggr.disabled.add(item.key);
          }
          if (item.props.keepMounted && item.key !== null) {
            aggr.keepMounted.add(item.key);
          }
          return aggr;
        },
        {
          disabled: /* @__PURE__ */ new Set(),
          keepMounted: /* @__PURE__ */ new Set(),
          consumerIds: {}
        }
      );
    }, [childrenArray]);
    const [selectedIndex, setSelectedIndex] = (0, import_useControllableState.useControllableState)({
      defaultValue: defaultIndexProp ?? 0,
      value: selectedIndexProp,
      onChange: onChangeProp
    });
    const ref = (0, import_react.useRef)(null);
    (0, import_react.useLayoutEffect)(() => {
      const tabList = ref.current;
      if (!tabList) {
        return;
      }
      const selectedTab = tabList.querySelector(
        '[aria-selected="true"]'
      );
      if (!selectedTab) {
        return;
      }
      const tabListRect = tabList.getBoundingClientRect();
      const tabRect = selectedTab.getBoundingClientRect();
      const leftArrow = tabList.previousElementSibling;
      const rightArrow = tabList.nextElementSibling;
      const leftOffset = leftArrow?.offsetWidth ?? 0;
      const rightOffset = rightArrow?.offsetWidth ?? 0;
      const ADJACENT_TAB_PEEK_WIDTH_PX = 24;
      const isHiddenLeft = tabRect.left < tabListRect.left + leftOffset;
      const isHiddenRight = tabRect.right > tabListRect.right - rightOffset;
      if (isHiddenLeft) {
        const targetLeftPosition = tabListRect.left + leftOffset + ADJACENT_TAB_PEEK_WIDTH_PX;
        const scrollAmount = targetLeftPosition - tabRect.left;
        tabList.scrollLeft -= scrollAmount;
      } else if (isHiddenRight) {
        const targetRightPosition = tabListRect.right - rightOffset - ADJACENT_TAB_PEEK_WIDTH_PX;
        const scrollAmount = tabRect.right - targetRightPosition;
        tabList.scrollLeft += scrollAmount;
      }
    }, [selectedIndex]);
    const { containerRef } = (0, import_useRovingFocus.useRovingFocus)({
      direction: "both",
      focusableChildrenSelector: '[type="button"],[role="tab"]',
      initialIndex: selectedIndex
    });
    const rootRef = (0, import_useMergeRefs.useMergeRefs)([ref, containerRef]);
    const [isOverflowingLeft, setIsOverflowingLeft] = (0, import_react.useState)(false);
    const [isOverflowingRight, setIsOverflowingRight] = (0, import_react.useState)(false);
    const slideTabsLeft = () => {
      ref.current?.scrollBy({
        left: -ref.current.offsetWidth / 2,
        behavior: "smooth"
      });
    };
    const slideTabsRight = () => {
      ref.current?.scrollBy({
        left: ref.current.offsetWidth / 2,
        behavior: "smooth"
      });
    };
    const updateOverflow = (0, import_react.useCallback)(() => {
      if (ref.current) {
        const tabsWidth = ref.current.offsetWidth;
        const tabsScrollWidth = ref.current.scrollWidth;
        const scrollDiff = tabsScrollWidth - tabsWidth;
        const tabsScrollLeft = Math.ceil(ref.current.scrollLeft);
        setIsOverflowingRight(
          tabsScrollWidth > tabsWidth && tabsScrollLeft < scrollDiff
        );
        setIsOverflowingLeft(
          tabsScrollWidth > tabsWidth && tabsScrollLeft !== 0
        );
      }
    }, []);
    const { width: containerWidth, height: containerHeight } = (0, import_use_resize_observer.default)({
      ref: ref.current
    });
    (0, import_react.useLayoutEffect)(() => {
      updateOverflow();
    }, [containerWidth, containerHeight, updateOverflow]);
    const getSelectedTabContent = (key, childrenArray2) => {
      const tab = childrenArray2.find((child) => child.key === key);
      if (tab?.props?.children) {
        return tab.props.children;
      }
      return null;
    };
    const getTabPanelId = (key, options) => {
      if (!options || options.keepMounted) {
        return key !== null ? `${key}-tabPanel` : void 0;
      }
      return key !== null && options.isSelected ? `${key}-tabPanel` : void 0;
    };
    const MemoizedSelectedTabPanel = (0, import_react.useMemo)(
      () => childrenArray.map((child, index) => {
        const isSelected = index === selectedIndex;
        const childKey = child.key;
        if (childKey !== null && (tabsInfo.keepMounted.has(childKey) || isSelected)) {
          return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_TabPanel.TabPanel,
            {
              style: {
                display: !isSelected ? "none" : ""
              },
              id: getTabPanelId(childKey),
              disabled: disabledTabsProp || tabsInfo.disabled.has(childKey),
              panelOverflow,
              "aria-labelledby": tabsInfo.consumerIds[childKey] || childKey,
              children: getSelectedTabContent(childKey, childrenArray)
            },
            childKey
          );
        }
        return null;
      }),
      [childrenArray, disabledTabsProp, panelOverflow, selectedIndex, tabsInfo]
    );
    const handleKeyDown = (0, import_react.useCallback)(
      (event) => {
        const selectNextTab = () => {
          setSelectedIndex(
            (curr) => curr + 1 > childrenArray.length - 1 ? 0 : curr + 1
          );
        };
        const selectPrevTab = () => {
          setSelectedIndex(
            (curr) => curr - 1 < 0 ? childrenArray.length - 1 : curr - 1
          );
        };
        if (["ArrowRight", "ArrowDown"].includes(event.code)) {
          selectNextTab();
        }
        if (["ArrowLeft", "ArrowUp"].includes(event.code)) {
          selectPrevTab();
        }
      },
      [setSelectedIndex, childrenArray.length]
    );
    const MemoizedTabsList = (0, import_react.useMemo)(() => {
      return childrenArray.map((item, index) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_RenderTab.RenderTab,
        {
          tabItem: item,
          isSelected: selectedIndex === index && !item.props.disabled && !disabledTabsProp,
          isDisabled: item.props.disabled || disabledTabsProp,
          onActivate: () => {
            setSelectedIndex(index);
          },
          onKeyDown: handleKeyDown,
          ariaControls: getTabPanelId(item.key, {
            keepMounted: item.props.keepMounted,
            isSelected: selectedIndex === index && !item.props.disabled && !disabledTabsProp
          })
        },
        item.key
      ));
    }, [
      childrenArray,
      selectedIndex,
      disabledTabsProp,
      setSelectedIndex,
      handleKeyDown
    ]);
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      "div",
      {
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        ref: forwardedRef,
        className: (0, import_clsx.default)(import_Tabs_css.tabsCSS, consumerClassName),
        style: consumerStyle,
        ...restProps,
        ...behavioralTrackingProps,
        "data-dt-component": COMPONENT_NAME,
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { className: import_Tabs_css.tabsArrowWrapperCSS, children: [
            isOverflowingLeft && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_Button.Button,
              {
                className: import_Tabs_css.tabsArrowButtonCSS,
                style: { left: 0 },
                variant: "default",
                onClick: slideTabsLeft,
                "aria-label": intl.formatMessage({
                  defaultMessage: "scroll tabs left",
                  id: "Cv2AP7VKybeuopCP",
                  description: "Aria label to scroll the tabs to the left"
                }),
                "aria-controls": tabsListId,
                children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ChevronLeftIcon, {}) })
              }
            ),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              "div",
              {
                className: (0, import_clsx.default)(import_Tabs_css.tabsWrapperCSS, {
                  [import_Tabs_css.tabsWrapperOverflowCSS.left]: isOverflowingLeft && !isOverflowingRight,
                  [import_Tabs_css.tabsWrapperOverflowCSS.right]: !isOverflowingLeft && isOverflowingRight,
                  [import_Tabs_css.tabsWrapperOverflowCSS.both]: isOverflowingLeft && isOverflowingRight
                }),
                role: "tablist",
                "aria-orientation": "horizontal",
                id: tabsListId,
                ref: rootRef,
                tabIndex: -1,
                onScroll: updateOverflow,
                children: MemoizedTabsList
              }
            ),
            isOverflowingRight && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_Button.Button,
              {
                className: import_Tabs_css.tabsArrowButtonCSS,
                style: { right: 0 },
                onClick: slideTabsRight,
                "aria-label": intl.formatMessage({
                  defaultMessage: "scroll tabs right",
                  id: "2YBMATuaXx9/4zBP",
                  description: "Aria label to scroll the tabs to the right"
                }),
                "aria-controls": tabsListId,
                children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Suffix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ChevronRightIcon, {}) })
              }
            )
          ] }),
          MemoizedSelectedTabPanel
        ]
      }
    );
  }
);
Tabs.displayName = "Tabs";
