// @dynatrace/strato-components/esm/charts/xy-chart/components/thresholds/builder.js
import { getYValue } from "./get-y-value.js";
import { isThresholdDynamic } from "./type-guards.js";
import { axesPositionsIncludes, axisPositionIncludes } from "./utils.js";
import { THRESHOLD_BAR_THICKNESS } from "../../../core/components/threshold-bar/types/constants.js";
import { isPointThreshold } from "../../../core/types/type-guards.js";
import { getValidColor } from "../../../core/utils/getValidColor.js";
import { NoThresholdData } from "../../context/XYChartThresholds.context.js";
import {
  castDynamicThresholdsDataType,
  castStaticThresholdsDataType
} from "../../processing-strategies/utils/cast-data/cast-data-types.js";
import { isAnyNumericalAxis } from "../../types/guards.js";
function getStrokeOnly(threshold) {
  const firstData = isThresholdDynamic(threshold) ? threshold.data[0] : threshold.data;
  const firstMarkerDomain = {
    y0: getYValue(firstData, "min"),
    y1: getYValue(firstData, "max")
  };
  return (isPointThreshold(firstMarkerDomain.y0, firstMarkerDomain.y1) || threshold.strokeOnly) ?? false;
}
function hasValidAxes(threshold, yAxisMetadata, xAxisMetadata = void 0) {
  let xAxisValid = true;
  if (isThresholdDynamic(threshold) && xAxisMetadata) {
    xAxisValid = threshold.xAxisId !== void 0 && threshold.xAxisId in xAxisMetadata;
  }
  return xAxisValid && threshold.yAxisId in yAxisMetadata;
}
const buildThresholds = (yAxes, xAxes, thresholdSlots) => {
  const xAxisMetadata = xAxes.filter(isAnyNumericalAxis).reduce((acc, axis) => {
    acc[axis.id] = {
      position: axis.position,
      type: axis.type,
      reversed: axis.reversed
    };
    return acc;
  }, {});
  const yAxisMetadata = yAxes.filter(isAnyNumericalAxis).reduce((acc, axis) => {
    acc[axis.id] = {
      position: axis.position,
      type: axis.type
    };
    return acc;
  }, {});
  const leftAxisWithThresholds = /* @__PURE__ */ new Set();
  const rightAxisWithThresholds = /* @__PURE__ */ new Set();
  const thresholdsData = thresholdSlots.reduce(
    (acc, threshold) => {
      if (hasValidAxes(threshold, yAxisMetadata, xAxisMetadata)) {
        const yAxis = yAxisMetadata[threshold.yAxisId];
        const baseThresholdData = {
          id: crypto.randomUUID(),
          yAxisId: threshold.yAxisId,
          color: getValidColor(threshold.color),
          position: yAxis.position,
          label: threshold.label,
          strokeOnly: getStrokeOnly(threshold)
        };
        if (axisPositionIncludes("left")(yAxis.position)) {
          leftAxisWithThresholds.add(threshold.yAxisId);
        }
        if (axisPositionIncludes("right")(yAxis.position)) {
          rightAxisWithThresholds.add(threshold.yAxisId);
        }
        if (isThresholdDynamic(threshold)) {
          const xAxis = xAxisMetadata[threshold.xAxisId];
          const castedData = castDynamicThresholdsDataType(
            threshold.data,
            xAxis.type,
            yAxis.type
          );
          if (castedData.length) {
            const dynamicData = xAxis.reversed ? castedData.toReversed() : castedData;
            acc.push({
              ...baseThresholdData,
              xAxisId: threshold.xAxisId,
              data: dynamicData
            });
          }
        } else {
          const castedData = castStaticThresholdsDataType(
            threshold.data,
            yAxis.type
          );
          if (castedData) {
            acc.push({
              ...baseThresholdData,
              data: castedData
            });
          }
        }
      }
      return acc;
    },
    []
  );
  if (!thresholdsData.length) {
    return NoThresholdData;
  }
  const thresholds = {
    thresholds: thresholdsData,
    none: false,
    offset: {
      left: axesPositionsIncludes("left", thresholdsData) ? THRESHOLD_BAR_THICKNESS : 0,
      right: axesPositionsIncludes("right", thresholdsData) ? THRESHOLD_BAR_THICKNESS : 0
    },
    yAxes: {
      left: leftAxisWithThresholds,
      right: rightAxisWithThresholds
    }
  };
  return thresholds;
};
export {
  buildThresholds
};
