// @dynatrace/strato-components/charts/core/components/chart-layout/ChartLayout.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ChartLayout_exports = {};
__export(ChartLayout_exports, {
  ChartLayout: () => ChartLayout,
  ChartLayoutRoot: () => ChartLayoutRoot
});
module.exports = __toCommonJS(ChartLayout_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_LayoutStateHandler = require("./components/LayoutStateHandler.js");
var import_ResizableChartLayout = require("./components/ResizableChartLayout.js");
var import_ChartLayoutOrientation_context = require("./context/ChartLayoutOrientation.context.js");
var import_ChartLayoutRatio_context = require("./context/ChartLayoutRatio.context.js");
var import_ChartLayoutSize_context = require("./context/ChartLayoutSize.context.js");
var import_ChartLayoutTemplate_context = require("./context/ChartLayoutTemplate.context.js");
var import_useAutomaticRatio = require("./hooks/useAutomaticRatio.js");
var import_Graph = require("./slots/Graph.js");
var import_Legend = require("./slots/Legend.js");
var import_Provider = require("./slots/Provider.js");
var import_get_chart_layout_settings = require("./utils/get-chart-layout-settings.js");
var import_get_layout_grid_settings = require("./utils/get-layout-grid-settings.js");
var import_useBehavioralTrackingProps = require("../../../../core/hooks/useBehavioralTrackingProps.js");
var import_useMergeRefs = require("../../../../core/hooks/useMergeRefs.js");
var import_Grid = require("../../../../layouts/grid/Grid.js");
var import_DataDtrumContext = require("../../context/DataDtrumContext.js");
var import_useObserveContainerSize = require("../../hooks/useObserveContainerSize.js");
const ChartLayoutRoot = (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    chartHeight,
    chartWidth,
    children,
    isEmpty = false,
    emptyState,
    errorState,
    showLoader = false,
    className,
    style,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "data-testid": dataTestId = "chart-layout",
    "data-dt-component": dataDtComponent,
    "aria-label": ariaLabel,
    role
  } = props;
  const [behavioralTrackingProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(props);
  const [containerSizes, gridContainer] = (0, import_useObserveContainerSize.useObserveContainerSize)();
  const containerAvailableArea = {
    width: containerSizes.width,
    height: containerSizes.height,
    wrapperHeight: containerSizes.wrapperHeight
  };
  const chartLayoutRef = (0, import_react.useRef)(null);
  const mergedRefs = (0, import_useMergeRefs.useMergeRefs)([
    gridContainer,
    chartLayoutRef
  ]);
  const {
    graphArea,
    legendArea,
    legendSettings,
    provider: Provider2
  } = (0, import_get_chart_layout_settings.getChartLayoutSettings)(containerSizes, children, isEmpty);
  const {
    hidden,
    orientation,
    ratio,
    initialRatio,
    overrideRatio,
    onResize,
    showResizer,
    ratioBoundaries
  } = legendSettings;
  const internalRatioRef = (0, import_react.useRef)(void 0);
  if (internalRatioRef.current === void 0 && legendArea !== void 0) {
    internalRatioRef.current = initialRatio ?? ratio;
  }
  const legendRatio = internalRatioRef.current ?? ratio;
  const {
    legendCandidateSize,
    readyForAutoRatio,
    legendRef,
    refreshAutoRatio
  } = (0, import_useAutomaticRatio.useAutomaticRatio)(orientation, legendRatio, containerAvailableArea);
  const { layoutWidth, layoutHeight, templateAreas, gridTemplate } = (0, import_get_layout_grid_settings.getLayoutGridSettings)(
    chartWidth,
    chartHeight,
    containerSizes,
    !hidden,
    showResizer
  );
  (0, import_react.useImperativeHandle)(
    forwardedRef,
    () => ({
      element: chartLayoutRef?.current,
      refreshLegendSize: refreshAutoRatio
    })
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ChartLayoutTemplate_context.ChartLayoutTemplateContext.Provider, { value: { graphArea, legendArea }, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ChartLayoutOrientation_context.ChartLayoutOrientationContext.Provider, { value: { orientation }, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ChartLayoutRatio_context.ChartLayoutRatioContext.Provider,
    {
      value: {
        legendRef,
        ratio: legendRatio,
        ratioBoundaries,
        overrideRatio,
        readyForAutoRatio,
        legendCandidateSize
      },
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ChartLayoutSize_context.ChartLayoutSizeContext.Provider, { value: containerAvailableArea, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_DataDtrumContext.DataDtrumContext.Provider,
        {
          value: { dataDtrumAllow, dataDtrumMask },
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_Grid.Grid,
            {
              ...behavioralTrackingProps,
              "data-testid": dataTestId,
              "data-dtrum-mask": dataDtrumMask,
              "data-dtrum-allow": dataDtrumAllow,
              "data-dt-component": dataDtComponent,
              className,
              width: layoutWidth,
              height: layoutHeight,
              ref: mergedRefs,
              gridTemplateColumns: gridTemplate,
              gridTemplateRows: gridTemplate,
              gridTemplateAreas: templateAreas,
              gap: 0,
              "aria-label": ariaLabel,
              role,
              style: {
                position: "relative",
                overflow: "hidden",
                ...style
              },
              children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_LayoutStateHandler.LayoutStateHandler,
                {
                  empty: isEmpty,
                  height: layoutHeight,
                  emptyState,
                  errorState,
                  showLoader,
                  children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(Provider2, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                    import_ResizableChartLayout.ResizableChartLayout,
                    {
                      onResize,
                      showResizer,
                      showLegend: !hidden
                    }
                  ) })
                }
              )
            }
          )
        }
      ) })
    }
  ) }) });
});
ChartLayoutRoot.displayName = "ChartLayoutRoot";
const ChartLayout = Object.assign(ChartLayoutRoot, {
  Graph: import_Graph.Graph,
  Legend: import_Legend.Legend,
  Provider: import_Provider.Provider
});
