// @dynatrace-sdk/client-notification-v2/cjs/index.js
/**
 * @license
 * Copyright 2023-2026 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/client/notification-v2/src/index.ts
var src_exports = {};
__export(src_exports, {
  ApiClientError: () => import_shared_errors5.ApiClientError,
  ApiGatewayError: () => import_shared_errors5.ApiGatewayError,
  ClientRequestError: () => import_shared_errors5.ClientRequestError,
  DavisEventNameMatch: () => DavisEventNameMatch,
  DavisEventTriggerConfigType: () => DavisEventTriggerConfigType,
  DavisProblemTriggerConfigType: () => DavisProblemTriggerConfigType,
  EntityTagsMatch: () => EntityTagsMatch,
  ErrorEnvelopeError: () => ErrorEnvelopeError,
  EventNotificationsClient: () => EventNotificationsClient,
  EventQueryTriggerConfigType: () => EventQueryTriggerConfigType,
  EventType: () => EventType,
  InvalidResponseError: () => import_shared_errors5.InvalidResponseError,
  MaintenanceWindowTriggerBehaviorType: () => MaintenanceWindowTriggerBehaviorType,
  ResourceNotificationsClient: () => ResourceNotificationsClient,
  eventNotificationsClient: () => eventNotificationsClient,
  isApiClientError: () => import_shared_errors5.isApiClientError,
  isApiGatewayError: () => import_shared_errors5.isApiGatewayError,
  isClientRequestError: () => import_shared_errors5.isClientRequestError,
  isErrorEnvelopeError: () => isErrorEnvelopeError,
  isInvalidResponseError: () => import_shared_errors5.isInvalidResponseError,
  resourceNotificationsClient: () => resourceNotificationsClient
});
module.exports = __toCommonJS(src_exports);

// packages/client/notification-v2/src/lib/apis/event-notifications-api.ts
var import_http_client = require("@dynatrace-sdk/http-client");
var import_shared_errors3 = require("@dynatrace-sdk/shared-errors");

// packages/client/notification-v2/src/lib/error-envelopes/api-gateway-errors-handler.ts
var import_shared_errors = require("@dynatrace-sdk/shared-errors");
async function apiGatewayErrorsHandler(response) {
  if (response.headers?.["dynatrace-response-source"] === "API Gateway") {
    const responseBody = await response.body("json");
    throw new import_shared_errors.ApiGatewayError(response, responseBody);
  }
}

// packages/client/notification-v2/src/lib/error-envelopes/error-envelope-error.ts
var import_shared_errors2 = require("@dynatrace-sdk/shared-errors");
var ErrorEnvelopeError = class extends import_shared_errors2.ClientRequestError {
  isErrorEnvelopeError = true;
};
function isErrorEnvelopeError(e) {
  return e?.isApiClientError === true && e?.isClientRequestError === true && e?.isErrorEnvelopeError === true && e instanceof Error;
}

// packages/client/notification-v2/src/lib/error-envelopes/get-error-message.ts
function serializeData(data) {
  try {
    return JSON.stringify(data);
  } catch (e) {
    return String(data);
  }
}
function getMessagesFromErrorDetails(details) {
  const messages = [];
  Object.entries(details).forEach(([name, data]) => {
    if (!data)
      return;
    const serializedData = serializeData(data);
    switch (name) {
      case "missingScopes":
        messages.push(`Missing scopes: ${serializedData}`);
        break;
      default:
        messages.push(`${name}: ${serializedData}`);
    }
  }, []);
  return messages;
}
function getErrorMessage(errorBody, defaultMessage) {
  const error = errorBody;
  const msg = error?.error?.message || defaultMessage;
  const details = error?.error?.details || {};
  return [msg, ...getMessagesFromErrorDetails(details)].join(". ");
}

// packages/client/notification-v2/src/lib/utils/transformations.ts
var defaultRuleConfig = {
  force: false,
  datetime: false
};
function mergeKeys(key1, key2) {
  if (!key1)
    return key2;
  if (!key2)
    return key1;
  return `${key1}.${key2}`;
}
function matchKey(key, rules, cache) {
  if (cache.has(key)) {
    return cache.get(key);
  }
  for (const rule of rules) {
    const pattern = Array.isArray(rule) ? rule[0] : rule;
    const config = Array.isArray(rule) ? rule[1] : defaultRuleConfig;
    if (key === pattern) {
      const result = [pattern, config];
      cache.set(key, result);
      return result;
    }
    let i = 0;
    let j = 0;
    const keyLen = key.length;
    const patternLen = pattern.length;
    let isMatch = true;
    while (i < keyLen && j < patternLen) {
      let keyPartEnd = key.indexOf(".", i);
      let patternPartEnd = pattern.indexOf(".", j);
      if (keyPartEnd === -1)
        keyPartEnd = keyLen;
      if (patternPartEnd === -1)
        patternPartEnd = patternLen;
      const keyPart = key.slice(i, keyPartEnd);
      const patternPart = pattern.slice(j, patternPartEnd);
      if (patternPart !== "*" && patternPart !== "**" && keyPart !== patternPart) {
        isMatch = false;
        break;
      }
      i = keyPartEnd + 1;
      j = patternPartEnd + 1;
    }
    if (isMatch) {
      if (j < patternLen) {
        const remaining = pattern.slice(j);
        if (remaining !== "**" && remaining !== "*") {
          isMatch = false;
        }
      }
      if (i < keyLen && pattern.slice(j - 1) !== "**") {
        isMatch = false;
      }
    }
    if (isMatch) {
      const result = [pattern, config];
      cache.set(key, result);
      return result;
    }
  }
  cache.set(key, null);
  return null;
}
var potentialMatchForRoot = (pattern, key) => pattern.startsWith(key) || pattern.startsWith("*") || pattern.startsWith("**");
var potentialMatch = (pattern, key) => pattern.startsWith(key) || pattern.includes("*") || pattern.includes("**");
function partialMatch(key, keyPatterns) {
  const keyPatternsLen = keyPatterns.length;
  if (!key.includes(".")) {
    for (let i = 0; i < keyPatternsLen; i++) {
      const pattern = keyPatterns[i];
      if (potentialMatchForRoot(pattern, key))
        return true;
    }
    return false;
  }
  let dotIndex = -1;
  let currentKeyEnd = 0;
  while ((dotIndex = key.indexOf(".", currentKeyEnd)) !== -1) {
    const currentKey = key.slice(0, dotIndex);
    const isRoot = !currentKey.includes(".");
    for (let i = 0; i < keyPatternsLen; i++) {
      const pattern = keyPatterns[i];
      if (isRoot) {
        if (potentialMatchForRoot(pattern, currentKey))
          return true;
      } else if (potentialMatch(pattern, currentKey))
        return true;
    }
    currentKeyEnd = dotIndex + 1;
  }
  for (let i = 0; i < keyPatternsLen; i++) {
    const pattern = keyPatterns[i];
    if (potentialMatch(pattern, key))
      return true;
  }
  return false;
}
var dateTimePattern = /^(\d{4}-\d{2}-\d{2})(T\d{2}:\d{2}:\d{2}(?:\.\d+)?(Z|[+-]\d{2}:\d{2})?)?$/i;
function isDateLike(value) {
  if (typeof value === "string") {
    return dateTimePattern.test(value);
  }
  return false;
}
function transformValue(value, rules, direction) {
  const { force, datetime } = rules;
  if (direction === "to") {
    return force || isDateLike(value) ? new Date(value) : value;
  }
  if (value instanceof Date) {
    const iso = value.toISOString();
    return datetime ? iso : iso.split("T")[0];
  }
  return value;
}
function walk(obj, direction, rules, rulesCache, keyPatterns, currentKey = "") {
  if (rules.length <= 0) {
    return obj;
  }
  if (!partialMatch(currentKey, keyPatterns)) {
    return obj;
  }
  if (Array.isArray(obj)) {
    const match = matchKey(currentKey, rules, rulesCache);
    obj.forEach((item, idx) => {
      if (typeof item === "object" && item !== null) {
        walk(item, direction, rules, rulesCache, keyPatterns, currentKey);
      } else if (match) {
        obj[idx] = transformValue(obj[idx], match[1], direction);
      }
    });
  }
  if (typeof obj === "object" && obj !== null && !Array.isArray(obj)) {
    for (const key of Object.keys(obj)) {
      const mergedKey = mergeKeys(currentKey, key);
      const match = matchKey(mergedKey, rules, rulesCache);
      if (match) {
        obj[key] = transformValue(obj[key], match[1], direction);
      }
      walk(obj[key], direction, rules, rulesCache, keyPatterns, mergedKey);
    }
  }
  return obj;
}
function flattenKeyPatterns(keyPatterns) {
  return keyPatterns.map((k) => Array.isArray(k) ? k[0] : k);
}
function transform(direction, object, rules) {
  const flatKeyPatterns = flattenKeyPatterns(rules);
  const rulesCache = /* @__PURE__ */ new Map();
  return walk(object, direction, rules, rulesCache, flatKeyPatterns);
}
function transformRequest(object, rules) {
  return transform("from", object, rules);
}
function transformResponse(object, rules) {
  return transform("to", object, rules);
}

// packages/client/notification-v2/src/lib/utils/url-helpers.ts
var encodeQueryParam = (key, value) => {
  const encodedKey = encodeURIComponent(key);
  return `${encodedKey}=${encodeURIComponent(typeof value === "number" ? value : String(value))}`;
};
var addExplodedArrayQueryParam = (query, key) => {
  const arrayValue = query[key];
  return arrayValue.map((value) => encodeQueryParam(key, value)).join("&");
};
var addNonExplodedArrayQueryParams = (query, key) => {
  const encodedKey = encodeURIComponent(key);
  const encodedParamsList = query[key].map((value) => encodeURIComponent(typeof value === "number" ? value : String(value))).join(",");
  return `${encodedKey}=${encodedParamsList}`;
};
var addQueryParam = (query, key) => encodeQueryParam(key, query[key]);
var arrayQueryParams = (query, key, explode) => {
  return explode ? addExplodedArrayQueryParam(query, key) : addNonExplodedArrayQueryParams(query, key);
};
var toQueryString = (rawQuery, flags = {}) => {
  const query = rawQuery || {};
  const keys = Object.keys(query).filter((key) => typeof query[key] !== "undefined");
  const queryString = keys.map(
    (key) => Array.isArray(query[key]) ? arrayQueryParams(query, key, flags.explode?.hasOwnProperty(key) ? flags.explode[key] : true) : addQueryParam(query, key)
  ).join("&");
  return queryString ? `?${queryString}` : "";
};

// packages/client/notification-v2/src/lib/apis/event-notifications-api.ts
var EventNotificationsClient = class {
  httpClient;
  shouldTransformDates = true;
  /**
   * @param {HttpClient} httpClientImplementation - You can provide custom http client as a parameter to constructor of a EventNotificationsClient. Custom http client should implement {HttpClient} interface.
   * @param {ApiClientOptions} [options] - Optional configuration for SDK Client.
   * @example
   * const eventNotificationsClientNodeJs = new EventNotificationsClient(yourCustomImplementation);
   */
  constructor(httpClientImplementation, options) {
    this.httpClient = httpClientImplementation;
    this.shouldTransformDates = options?.transformDates ?? true;
  }
  /**
   * **Required scope:** notification:notifications:read
   *
   * @returns {Promise<PaginatedEventNotificationList>}
   *
   * @example <caption>Code example</caption>
   * import { eventNotificationsClient } from "@dynatrace-sdk/client-notification-v2";
   *
   * const data =
   *   await eventNotificationsClient.getEventNotifications();
   *
   * @throws {ErrorEnvelopeError}
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async getEventNotifications(config = {}) {
    if (!config) {
      throw new import_shared_errors3.ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const query = toQueryString({
      appId: config.appId,
      limit: config.limit,
      notificationType: config.notificationType,
      offset: config.offset,
      ordering: config.ordering,
      owner: config.owner,
      resourceId: config.resourceId,
      search: config.search
    });
    try {
      const response = await this.httpClient.send({
        url: `/platform/notification/v2/event-notifications${query}`,
        method: "GET",
        headers: {
          Accept: "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? transformResponse(responseValue, [
          ["results.modificationInfo.createdTime", { datetime: true }],
          ["results.modificationInfo.lastModifiedTime", { datetime: true }]
        ]) : responseValue;
      } catch (err) {
        throw new import_shared_errors3.InvalidResponseError(
          `EventNotificationsClient.getEventNotifications:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if ((0, import_http_client.isHttpClientAbortError)(e) || (0, import_shared_errors3.isInvalidResponseError)(e)) {
        throw e;
      }
      if (!(0, import_http_client.isHttpClientResponseError)(e)) {
        throw new import_shared_errors3.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      if (response.status >= 400 && response.status <= 499) {
        const responseValue = await response.body("json");
        try {
          const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
          throw new ErrorEnvelopeError(`${response.status}`, response, errorBody, getErrorMessage(errorBody, ""), e);
        } catch (err) {
          if (err instanceof ErrorEnvelopeError) {
            throw err;
          }
          throw new import_shared_errors3.InvalidResponseError(
            `EventNotificationsClient.getEventNotifications:${response.status}`,
            err,
            responseValue,
            "ErrorEnvelope",
            void 0,
            response
          );
        }
      } else if (response.status >= 500 && response.status <= 599) {
        const responseValue = await response.body("json");
        try {
          const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
          throw new ErrorEnvelopeError(`${response.status}`, response, errorBody, getErrorMessage(errorBody, ""), e);
        } catch (err) {
          if (err instanceof ErrorEnvelopeError) {
            throw err;
          }
          throw new import_shared_errors3.InvalidResponseError(
            `EventNotificationsClient.getEventNotifications:${response.status}`,
            err,
            responseValue,
            "ErrorEnvelope",
            void 0,
            response
          );
        }
      } else {
        try {
          const responseValue = await response.body("json");
          throw new import_shared_errors3.ClientRequestError(
            `${response.status}`,
            response,
            responseValue,
            getErrorMessage(responseValue, `Unexpected api response: code=${response.status} body="${responseValue}"`)
          );
        } catch (error) {
          if (error instanceof import_shared_errors3.ClientRequestError) {
            throw error;
          }
          const responseValue = await response.body("text").catch(() => "");
          throw new import_shared_errors3.InvalidResponseError(
            `${response.status}`,
            error,
            responseValue,
            "json",
            "Unexpected api response",
            response
          );
        }
      }
    }
  }
  /**
   * **Required scope:** notification:notifications:write
   *
   * @returns {Promise<EventNotification>}
   *
   * @example <caption>Code example</caption>
   * import { eventNotificationsClient } from "@dynatrace-sdk/client-notification-v2";
   *
   * const data =
   *   await eventNotificationsClient.createEventNotification({
   *     body: {
   *       resourceId: "...",
   *       notificationType: "...",
   *       triggerConfiguration: {
   *         type: "event",
   *         value: { query: "..." },
   *       },
   *     },
   *   });
   *
   * @throws {ErrorEnvelopeError}
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async createEventNotification(config) {
    if (!config) {
      throw new import_shared_errors3.ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const encodedBody = this.shouldTransformDates ? transformRequest(config.body, []) : config.body;
    try {
      const response = await this.httpClient.send({
        url: `/platform/notification/v2/event-notifications`,
        method: "POST",
        requestBodyType: "json",
        body: encodedBody,
        headers: {
          "Content-Type": "application/json",
          Accept: "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [201].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? transformResponse(responseValue, [
          ["modificationInfo.createdTime", { datetime: true }],
          ["modificationInfo.lastModifiedTime", { datetime: true }]
        ]) : responseValue;
      } catch (err) {
        throw new import_shared_errors3.InvalidResponseError(
          `EventNotificationsClient.createEventNotification:201`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if ((0, import_http_client.isHttpClientAbortError)(e) || (0, import_shared_errors3.isInvalidResponseError)(e)) {
        throw e;
      }
      if (!(0, import_http_client.isHttpClientResponseError)(e)) {
        throw new import_shared_errors3.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      if (response.status >= 400 && response.status <= 499) {
        const responseValue = await response.body("json");
        try {
          const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
          throw new ErrorEnvelopeError(`${response.status}`, response, errorBody, getErrorMessage(errorBody, ""), e);
        } catch (err) {
          if (err instanceof ErrorEnvelopeError) {
            throw err;
          }
          throw new import_shared_errors3.InvalidResponseError(
            `EventNotificationsClient.createEventNotification:${response.status}`,
            err,
            responseValue,
            "ErrorEnvelope",
            void 0,
            response
          );
        }
      } else if (response.status >= 500 && response.status <= 599) {
        const responseValue = await response.body("json");
        try {
          const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
          throw new ErrorEnvelopeError(`${response.status}`, response, errorBody, getErrorMessage(errorBody, ""), e);
        } catch (err) {
          if (err instanceof ErrorEnvelopeError) {
            throw err;
          }
          throw new import_shared_errors3.InvalidResponseError(
            `EventNotificationsClient.createEventNotification:${response.status}`,
            err,
            responseValue,
            "ErrorEnvelope",
            void 0,
            response
          );
        }
      } else {
        try {
          const responseValue = await response.body("json");
          throw new import_shared_errors3.ClientRequestError(
            `${response.status}`,
            response,
            responseValue,
            getErrorMessage(responseValue, `Unexpected api response: code=${response.status} body="${responseValue}"`)
          );
        } catch (error) {
          if (error instanceof import_shared_errors3.ClientRequestError) {
            throw error;
          }
          const responseValue = await response.body("text").catch(() => "");
          throw new import_shared_errors3.InvalidResponseError(
            `${response.status}`,
            error,
            responseValue,
            "json",
            "Unexpected api response",
            response
          );
        }
      }
    }
  }
  /**
   * **Required scope:** notification:notifications:read
   *
   * @returns {Promise<EventNotification>}
   *
   * @example <caption>Code example</caption>
   * import { eventNotificationsClient } from "@dynatrace-sdk/client-notification-v2";
   *
   * const data =
   *   await eventNotificationsClient.getEventNotification({
   *     id: "...",
   *   });
   *
   * @throws {ErrorEnvelopeError}
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async getEventNotification(config) {
    try {
      const response = await this.httpClient.send({
        url: `/platform/notification/v2/event-notifications/${config.id}`,
        method: "GET",
        headers: {
          Accept: "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? transformResponse(responseValue, [
          ["modificationInfo.createdTime", { datetime: true }],
          ["modificationInfo.lastModifiedTime", { datetime: true }]
        ]) : responseValue;
      } catch (err) {
        throw new import_shared_errors3.InvalidResponseError(
          `EventNotificationsClient.getEventNotification:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if ((0, import_http_client.isHttpClientAbortError)(e) || (0, import_shared_errors3.isInvalidResponseError)(e)) {
        throw e;
      }
      if (!(0, import_http_client.isHttpClientResponseError)(e)) {
        throw new import_shared_errors3.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      if (response.status >= 400 && response.status <= 499) {
        const responseValue = await response.body("json");
        try {
          const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
          throw new ErrorEnvelopeError(`${response.status}`, response, errorBody, getErrorMessage(errorBody, ""), e);
        } catch (err) {
          if (err instanceof ErrorEnvelopeError) {
            throw err;
          }
          throw new import_shared_errors3.InvalidResponseError(
            `EventNotificationsClient.getEventNotification:${response.status}`,
            err,
            responseValue,
            "ErrorEnvelope",
            void 0,
            response
          );
        }
      } else if (response.status >= 500 && response.status <= 599) {
        const responseValue = await response.body("json");
        try {
          const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
          throw new ErrorEnvelopeError(`${response.status}`, response, errorBody, getErrorMessage(errorBody, ""), e);
        } catch (err) {
          if (err instanceof ErrorEnvelopeError) {
            throw err;
          }
          throw new import_shared_errors3.InvalidResponseError(
            `EventNotificationsClient.getEventNotification:${response.status}`,
            err,
            responseValue,
            "ErrorEnvelope",
            void 0,
            response
          );
        }
      } else {
        try {
          const responseValue = await response.body("json");
          throw new import_shared_errors3.ClientRequestError(
            `${response.status}`,
            response,
            responseValue,
            getErrorMessage(responseValue, `Unexpected api response: code=${response.status} body="${responseValue}"`)
          );
        } catch (error) {
          if (error instanceof import_shared_errors3.ClientRequestError) {
            throw error;
          }
          const responseValue = await response.body("text").catch(() => "");
          throw new import_shared_errors3.InvalidResponseError(
            `${response.status}`,
            error,
            responseValue,
            "json",
            "Unexpected api response",
            response
          );
        }
      }
    }
  }
  /**
   * **Required scope:** notification:notifications:write
   *
   * @returns {Promise<EventNotification>}
   *
   * @example <caption>Code example</caption>
   * import { eventNotificationsClient } from "@dynatrace-sdk/client-notification-v2";
   *
   * const data =
   *   await eventNotificationsClient.updateEventNotification({
   *     id: "...",
   *     body: {
   *       triggerConfiguration: {
   *         type: "event",
   *         value: { query: "..." },
   *       },
   *     },
   *   });
   *
   * @throws {ErrorEnvelopeError}
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async updateEventNotification(config) {
    if (!config) {
      throw new import_shared_errors3.ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const encodedBody = this.shouldTransformDates ? transformRequest(config.body, []) : config.body;
    try {
      const response = await this.httpClient.send({
        url: `/platform/notification/v2/event-notifications/${config.id}`,
        method: "PUT",
        requestBodyType: "json",
        body: encodedBody,
        headers: {
          "Content-Type": "application/json",
          Accept: "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? transformResponse(responseValue, [
          ["modificationInfo.createdTime", { datetime: true }],
          ["modificationInfo.lastModifiedTime", { datetime: true }]
        ]) : responseValue;
      } catch (err) {
        throw new import_shared_errors3.InvalidResponseError(
          `EventNotificationsClient.updateEventNotification:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if ((0, import_http_client.isHttpClientAbortError)(e) || (0, import_shared_errors3.isInvalidResponseError)(e)) {
        throw e;
      }
      if (!(0, import_http_client.isHttpClientResponseError)(e)) {
        throw new import_shared_errors3.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      if (response.status >= 400 && response.status <= 499) {
        const responseValue = await response.body("json");
        try {
          const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
          throw new ErrorEnvelopeError(`${response.status}`, response, errorBody, getErrorMessage(errorBody, ""), e);
        } catch (err) {
          if (err instanceof ErrorEnvelopeError) {
            throw err;
          }
          throw new import_shared_errors3.InvalidResponseError(
            `EventNotificationsClient.updateEventNotification:${response.status}`,
            err,
            responseValue,
            "ErrorEnvelope",
            void 0,
            response
          );
        }
      } else if (response.status >= 500 && response.status <= 599) {
        const responseValue = await response.body("json");
        try {
          const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
          throw new ErrorEnvelopeError(`${response.status}`, response, errorBody, getErrorMessage(errorBody, ""), e);
        } catch (err) {
          if (err instanceof ErrorEnvelopeError) {
            throw err;
          }
          throw new import_shared_errors3.InvalidResponseError(
            `EventNotificationsClient.updateEventNotification:${response.status}`,
            err,
            responseValue,
            "ErrorEnvelope",
            void 0,
            response
          );
        }
      } else {
        try {
          const responseValue = await response.body("json");
          throw new import_shared_errors3.ClientRequestError(
            `${response.status}`,
            response,
            responseValue,
            getErrorMessage(responseValue, `Unexpected api response: code=${response.status} body="${responseValue}"`)
          );
        } catch (error) {
          if (error instanceof import_shared_errors3.ClientRequestError) {
            throw error;
          }
          const responseValue = await response.body("text").catch(() => "");
          throw new import_shared_errors3.InvalidResponseError(
            `${response.status}`,
            error,
            responseValue,
            "json",
            "Unexpected api response",
            response
          );
        }
      }
    }
  }
  /**
   * **Required scope:** notification:notifications:write
   *
   * @example <caption>Code example</caption>
   * import { eventNotificationsClient } from "@dynatrace-sdk/client-notification-v2";
   *
   * const data =
   *   await eventNotificationsClient.deleteEventNotification({
   *     id: "...",
   *   });
   *
   * @throws {ErrorEnvelopeError}
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async deleteEventNotification(config) {
    try {
      await this.httpClient.send({
        url: `/platform/notification/v2/event-notifications/${config.id}`,
        method: "DELETE",
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [204].includes(status);
        }
      });
      return;
    } catch (e) {
      if ((0, import_http_client.isHttpClientAbortError)(e)) {
        throw e;
      }
      if (!(0, import_http_client.isHttpClientResponseError)(e)) {
        throw new import_shared_errors3.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      if (response.status >= 400 && response.status <= 499) {
        const responseValue = await response.body("json");
        try {
          const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
          throw new ErrorEnvelopeError(`${response.status}`, response, errorBody, getErrorMessage(errorBody, ""), e);
        } catch (err) {
          if (err instanceof ErrorEnvelopeError) {
            throw err;
          }
          throw new import_shared_errors3.InvalidResponseError(
            `EventNotificationsClient.deleteEventNotification:${response.status}`,
            err,
            responseValue,
            "ErrorEnvelope",
            void 0,
            response
          );
        }
      } else if (response.status >= 500 && response.status <= 599) {
        const responseValue = await response.body("json");
        try {
          const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
          throw new ErrorEnvelopeError(`${response.status}`, response, errorBody, getErrorMessage(errorBody, ""), e);
        } catch (err) {
          if (err instanceof ErrorEnvelopeError) {
            throw err;
          }
          throw new import_shared_errors3.InvalidResponseError(
            `EventNotificationsClient.deleteEventNotification:${response.status}`,
            err,
            responseValue,
            "ErrorEnvelope",
            void 0,
            response
          );
        }
      } else {
        try {
          const responseValue = await response.body("json");
          throw new import_shared_errors3.ClientRequestError(
            `${response.status}`,
            response,
            responseValue,
            getErrorMessage(responseValue, `Unexpected api response: code=${response.status} body="${responseValue}"`)
          );
        } catch (error) {
          if (error instanceof import_shared_errors3.ClientRequestError) {
            throw error;
          }
          const responseValue = await response.body("text").catch(() => "");
          throw new import_shared_errors3.InvalidResponseError(
            `${response.status}`,
            error,
            responseValue,
            "json",
            "Unexpected api response",
            response
          );
        }
      }
    }
  }
  /**
   * **Required scope:** notification:notifications:write
   *
   * @returns {Promise<EventNotification>}
   *
   * @example <caption>Code example</caption>
   * import { eventNotificationsClient } from "@dynatrace-sdk/client-notification-v2";
   *
   * const data =
   *   await eventNotificationsClient.patchEventNotification({
   *     id: "...",
   *     body: {
   *       triggerConfiguration: {
   *         type: "event",
   *         value: { query: "..." },
   *       },
   *     },
   *   });
   *
   * @throws {ErrorEnvelopeError}
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async patchEventNotification(config) {
    if (!config) {
      throw new import_shared_errors3.ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const encodedBody = this.shouldTransformDates ? transformRequest(config.body, []) : config.body;
    try {
      const response = await this.httpClient.send({
        url: `/platform/notification/v2/event-notifications/${config.id}`,
        method: "PATCH",
        requestBodyType: "json",
        body: encodedBody,
        headers: {
          "Content-Type": "application/json",
          Accept: "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? transformResponse(responseValue, [
          ["modificationInfo.createdTime", { datetime: true }],
          ["modificationInfo.lastModifiedTime", { datetime: true }]
        ]) : responseValue;
      } catch (err) {
        throw new import_shared_errors3.InvalidResponseError(
          `EventNotificationsClient.patchEventNotification:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if ((0, import_http_client.isHttpClientAbortError)(e) || (0, import_shared_errors3.isInvalidResponseError)(e)) {
        throw e;
      }
      if (!(0, import_http_client.isHttpClientResponseError)(e)) {
        throw new import_shared_errors3.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      if (response.status >= 400 && response.status <= 499) {
        const responseValue = await response.body("json");
        try {
          const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
          throw new ErrorEnvelopeError(`${response.status}`, response, errorBody, getErrorMessage(errorBody, ""), e);
        } catch (err) {
          if (err instanceof ErrorEnvelopeError) {
            throw err;
          }
          throw new import_shared_errors3.InvalidResponseError(
            `EventNotificationsClient.patchEventNotification:${response.status}`,
            err,
            responseValue,
            "ErrorEnvelope",
            void 0,
            response
          );
        }
      } else if (response.status >= 500 && response.status <= 599) {
        const responseValue = await response.body("json");
        try {
          const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
          throw new ErrorEnvelopeError(`${response.status}`, response, errorBody, getErrorMessage(errorBody, ""), e);
        } catch (err) {
          if (err instanceof ErrorEnvelopeError) {
            throw err;
          }
          throw new import_shared_errors3.InvalidResponseError(
            `EventNotificationsClient.patchEventNotification:${response.status}`,
            err,
            responseValue,
            "ErrorEnvelope",
            void 0,
            response
          );
        }
      } else {
        try {
          const responseValue = await response.body("json");
          throw new import_shared_errors3.ClientRequestError(
            `${response.status}`,
            response,
            responseValue,
            getErrorMessage(responseValue, `Unexpected api response: code=${response.status} body="${responseValue}"`)
          );
        } catch (error) {
          if (error instanceof import_shared_errors3.ClientRequestError) {
            throw error;
          }
          const responseValue = await response.body("text").catch(() => "");
          throw new import_shared_errors3.InvalidResponseError(
            `${response.status}`,
            error,
            responseValue,
            "json",
            "Unexpected api response",
            response
          );
        }
      }
    }
  }
};
var eventNotificationsClient = /* @__PURE__ */ new EventNotificationsClient(import_http_client.httpClient);

// packages/client/notification-v2/src/lib/apis/resource-notifications-api.ts
var import_http_client2 = require("@dynatrace-sdk/http-client");
var import_shared_errors4 = require("@dynatrace-sdk/shared-errors");
var ResourceNotificationsClient = class {
  httpClient;
  shouldTransformDates = true;
  /**
   * @param {HttpClient} httpClientImplementation - You can provide custom http client as a parameter to constructor of a ResourceNotificationsClient. Custom http client should implement {HttpClient} interface.
   * @param {ApiClientOptions} [options] - Optional configuration for SDK Client.
   * @example
   * const resourceNotificationsClientNodeJs = new ResourceNotificationsClient(yourCustomImplementation);
   */
  constructor(httpClientImplementation, options) {
    this.httpClient = httpClientImplementation;
    this.shouldTransformDates = options?.transformDates ?? true;
  }
  /**
   * **Required scope:** notification:notifications:read
   *
   * @returns {Promise<PaginatedResourceNotificationList>}
   *
   * @example <caption>Code example</caption>
   * import { resourceNotificationsClient } from "@dynatrace-sdk/client-notification-v2";
   *
   * const data =
   *   await resourceNotificationsClient.getResourceNotifications();
   *
   * @throws {ErrorEnvelopeError}
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async getResourceNotifications(config = {}) {
    if (!config) {
      throw new import_shared_errors4.ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const query = toQueryString({
      appId: config.appId,
      limit: config.limit,
      notificationType: config.notificationType,
      offset: config.offset,
      ordering: config.ordering,
      resourceId: config.resourceId,
      search: config.search
    });
    try {
      const response = await this.httpClient.send({
        url: `/platform/notification/v2/resource-notifications${query}`,
        method: "GET",
        headers: {
          Accept: "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? transformResponse(responseValue, [
          ["results.modificationInfo.createdTime", { datetime: true }],
          ["results.modificationInfo.lastModifiedTime", { datetime: true }]
        ]) : responseValue;
      } catch (err) {
        throw new import_shared_errors4.InvalidResponseError(
          `ResourceNotificationsClient.getResourceNotifications:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if ((0, import_http_client2.isHttpClientAbortError)(e) || (0, import_shared_errors4.isInvalidResponseError)(e)) {
        throw e;
      }
      if (!(0, import_http_client2.isHttpClientResponseError)(e)) {
        throw new import_shared_errors4.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      if (response.status >= 400 && response.status <= 499) {
        const responseValue = await response.body("json");
        try {
          const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
          throw new ErrorEnvelopeError(`${response.status}`, response, errorBody, getErrorMessage(errorBody, ""), e);
        } catch (err) {
          if (err instanceof ErrorEnvelopeError) {
            throw err;
          }
          throw new import_shared_errors4.InvalidResponseError(
            `ResourceNotificationsClient.getResourceNotifications:${response.status}`,
            err,
            responseValue,
            "ErrorEnvelope",
            void 0,
            response
          );
        }
      } else if (response.status >= 500 && response.status <= 599) {
        const responseValue = await response.body("json");
        try {
          const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
          throw new ErrorEnvelopeError(`${response.status}`, response, errorBody, getErrorMessage(errorBody, ""), e);
        } catch (err) {
          if (err instanceof ErrorEnvelopeError) {
            throw err;
          }
          throw new import_shared_errors4.InvalidResponseError(
            `ResourceNotificationsClient.getResourceNotifications:${response.status}`,
            err,
            responseValue,
            "ErrorEnvelope",
            void 0,
            response
          );
        }
      } else {
        try {
          const responseValue = await response.body("json");
          throw new import_shared_errors4.ClientRequestError(
            `${response.status}`,
            response,
            responseValue,
            getErrorMessage(responseValue, `Unexpected api response: code=${response.status} body="${responseValue}"`)
          );
        } catch (error) {
          if (error instanceof import_shared_errors4.ClientRequestError) {
            throw error;
          }
          const responseValue = await response.body("text").catch(() => "");
          throw new import_shared_errors4.InvalidResponseError(
            `${response.status}`,
            error,
            responseValue,
            "json",
            "Unexpected api response",
            response
          );
        }
      }
    }
  }
  /**
   * **Required scope:** notification:notifications:write
   *
   * @returns {Promise<ResourceNotification>}
   *
   * @example <caption>Code example</caption>
   * import { resourceNotificationsClient } from "@dynatrace-sdk/client-notification-v2";
   *
   * const data =
   *   await resourceNotificationsClient.createResourceNotification(
   *     {
   *       body: { notificationType: "...", resourceId: "..." },
   *     },
   *   );
   *
   * @throws {ErrorEnvelopeError}
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async createResourceNotification(config) {
    if (!config) {
      throw new import_shared_errors4.ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const encodedBody = this.shouldTransformDates ? transformRequest(config.body, []) : config.body;
    try {
      const response = await this.httpClient.send({
        url: `/platform/notification/v2/resource-notifications`,
        method: "POST",
        requestBodyType: "json",
        body: encodedBody,
        headers: {
          "Content-Type": "application/json",
          Accept: "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [201].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? transformResponse(responseValue, [
          ["modificationInfo.createdTime", { datetime: true }],
          ["modificationInfo.lastModifiedTime", { datetime: true }]
        ]) : responseValue;
      } catch (err) {
        throw new import_shared_errors4.InvalidResponseError(
          `ResourceNotificationsClient.createResourceNotification:201`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if ((0, import_http_client2.isHttpClientAbortError)(e) || (0, import_shared_errors4.isInvalidResponseError)(e)) {
        throw e;
      }
      if (!(0, import_http_client2.isHttpClientResponseError)(e)) {
        throw new import_shared_errors4.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      if (response.status >= 400 && response.status <= 499) {
        const responseValue = await response.body("json");
        try {
          const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
          throw new ErrorEnvelopeError(`${response.status}`, response, errorBody, getErrorMessage(errorBody, ""), e);
        } catch (err) {
          if (err instanceof ErrorEnvelopeError) {
            throw err;
          }
          throw new import_shared_errors4.InvalidResponseError(
            `ResourceNotificationsClient.createResourceNotification:${response.status}`,
            err,
            responseValue,
            "ErrorEnvelope",
            void 0,
            response
          );
        }
      } else if (response.status >= 500 && response.status <= 599) {
        const responseValue = await response.body("json");
        try {
          const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
          throw new ErrorEnvelopeError(`${response.status}`, response, errorBody, getErrorMessage(errorBody, ""), e);
        } catch (err) {
          if (err instanceof ErrorEnvelopeError) {
            throw err;
          }
          throw new import_shared_errors4.InvalidResponseError(
            `ResourceNotificationsClient.createResourceNotification:${response.status}`,
            err,
            responseValue,
            "ErrorEnvelope",
            void 0,
            response
          );
        }
      } else {
        try {
          const responseValue = await response.body("json");
          throw new import_shared_errors4.ClientRequestError(
            `${response.status}`,
            response,
            responseValue,
            getErrorMessage(responseValue, `Unexpected api response: code=${response.status} body="${responseValue}"`)
          );
        } catch (error) {
          if (error instanceof import_shared_errors4.ClientRequestError) {
            throw error;
          }
          const responseValue = await response.body("text").catch(() => "");
          throw new import_shared_errors4.InvalidResponseError(
            `${response.status}`,
            error,
            responseValue,
            "json",
            "Unexpected api response",
            response
          );
        }
      }
    }
  }
  /**
   * **Required scope:** notification:notifications:write
   *
   * @example <caption>Code example</caption>
   * import { resourceNotificationsClient } from "@dynatrace-sdk/client-notification-v2";
   *
   * const data =
   *   await resourceNotificationsClient.deleteResourceNotificationByTypeAndResource(
   *     { notificationType: "...", resourceId: "..." },
   *   );
   *
   * @throws {ErrorEnvelopeError}
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async deleteResourceNotificationByTypeAndResource(config) {
    if (!config) {
      throw new import_shared_errors4.ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const query = toQueryString({ appId: config.appId });
    try {
      await this.httpClient.send({
        url: `/platform/notification/v2/resource-notifications/${config.notificationType}/${config.resourceId}${query}`,
        method: "DELETE",
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [204].includes(status);
        }
      });
      return;
    } catch (e) {
      if ((0, import_http_client2.isHttpClientAbortError)(e)) {
        throw e;
      }
      if (!(0, import_http_client2.isHttpClientResponseError)(e)) {
        throw new import_shared_errors4.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      if (response.status >= 400 && response.status <= 499) {
        const responseValue = await response.body("json");
        try {
          const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
          throw new ErrorEnvelopeError(`${response.status}`, response, errorBody, getErrorMessage(errorBody, ""), e);
        } catch (err) {
          if (err instanceof ErrorEnvelopeError) {
            throw err;
          }
          throw new import_shared_errors4.InvalidResponseError(
            `ResourceNotificationsClient.deleteResourceNotificationByTypeAndResource:${response.status}`,
            err,
            responseValue,
            "ErrorEnvelope",
            void 0,
            response
          );
        }
      } else if (response.status >= 500 && response.status <= 599) {
        const responseValue = await response.body("json");
        try {
          const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
          throw new ErrorEnvelopeError(`${response.status}`, response, errorBody, getErrorMessage(errorBody, ""), e);
        } catch (err) {
          if (err instanceof ErrorEnvelopeError) {
            throw err;
          }
          throw new import_shared_errors4.InvalidResponseError(
            `ResourceNotificationsClient.deleteResourceNotificationByTypeAndResource:${response.status}`,
            err,
            responseValue,
            "ErrorEnvelope",
            void 0,
            response
          );
        }
      } else {
        try {
          const responseValue = await response.body("json");
          throw new import_shared_errors4.ClientRequestError(
            `${response.status}`,
            response,
            responseValue,
            getErrorMessage(responseValue, `Unexpected api response: code=${response.status} body="${responseValue}"`)
          );
        } catch (error) {
          if (error instanceof import_shared_errors4.ClientRequestError) {
            throw error;
          }
          const responseValue = await response.body("text").catch(() => "");
          throw new import_shared_errors4.InvalidResponseError(
            `${response.status}`,
            error,
            responseValue,
            "json",
            "Unexpected api response",
            response
          );
        }
      }
    }
  }
  /**
   * **Required scope:** notification:notifications:read
   *
   * @returns {Promise<ResourceNotification>}
   *
   * @example <caption>Code example</caption>
   * import { resourceNotificationsClient } from "@dynatrace-sdk/client-notification-v2";
   *
   * const data =
   *   await resourceNotificationsClient.getResourceNotification(
   *     { id: "..." },
   *   );
   *
   * @throws {ErrorEnvelopeError}
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async getResourceNotification(config) {
    if (!config) {
      throw new import_shared_errors4.ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const query = toQueryString({ appId: config.appId });
    try {
      const response = await this.httpClient.send({
        url: `/platform/notification/v2/resource-notifications/${config.id}${query}`,
        method: "GET",
        headers: {
          Accept: "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? transformResponse(responseValue, [
          ["modificationInfo.createdTime", { datetime: true }],
          ["modificationInfo.lastModifiedTime", { datetime: true }]
        ]) : responseValue;
      } catch (err) {
        throw new import_shared_errors4.InvalidResponseError(
          `ResourceNotificationsClient.getResourceNotification:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if ((0, import_http_client2.isHttpClientAbortError)(e) || (0, import_shared_errors4.isInvalidResponseError)(e)) {
        throw e;
      }
      if (!(0, import_http_client2.isHttpClientResponseError)(e)) {
        throw new import_shared_errors4.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      if (response.status >= 400 && response.status <= 499) {
        const responseValue = await response.body("json");
        try {
          const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
          throw new ErrorEnvelopeError(`${response.status}`, response, errorBody, getErrorMessage(errorBody, ""), e);
        } catch (err) {
          if (err instanceof ErrorEnvelopeError) {
            throw err;
          }
          throw new import_shared_errors4.InvalidResponseError(
            `ResourceNotificationsClient.getResourceNotification:${response.status}`,
            err,
            responseValue,
            "ErrorEnvelope",
            void 0,
            response
          );
        }
      } else if (response.status >= 500 && response.status <= 599) {
        const responseValue = await response.body("json");
        try {
          const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
          throw new ErrorEnvelopeError(`${response.status}`, response, errorBody, getErrorMessage(errorBody, ""), e);
        } catch (err) {
          if (err instanceof ErrorEnvelopeError) {
            throw err;
          }
          throw new import_shared_errors4.InvalidResponseError(
            `ResourceNotificationsClient.getResourceNotification:${response.status}`,
            err,
            responseValue,
            "ErrorEnvelope",
            void 0,
            response
          );
        }
      } else {
        try {
          const responseValue = await response.body("json");
          throw new import_shared_errors4.ClientRequestError(
            `${response.status}`,
            response,
            responseValue,
            getErrorMessage(responseValue, `Unexpected api response: code=${response.status} body="${responseValue}"`)
          );
        } catch (error) {
          if (error instanceof import_shared_errors4.ClientRequestError) {
            throw error;
          }
          const responseValue = await response.body("text").catch(() => "");
          throw new import_shared_errors4.InvalidResponseError(
            `${response.status}`,
            error,
            responseValue,
            "json",
            "Unexpected api response",
            response
          );
        }
      }
    }
  }
  /**
   * **Required scope:** notification:notifications:write
   *
   * @example <caption>Code example</caption>
   * import { resourceNotificationsClient } from "@dynatrace-sdk/client-notification-v2";
   *
   * const data =
   *   await resourceNotificationsClient.deleteResourceNotification(
   *     { id: "..." },
   *   );
   *
   * @throws {ErrorEnvelopeError}
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async deleteResourceNotification(config) {
    if (!config) {
      throw new import_shared_errors4.ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const query = toQueryString({ appId: config.appId });
    try {
      await this.httpClient.send({
        url: `/platform/notification/v2/resource-notifications/${config.id}${query}`,
        method: "DELETE",
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [204].includes(status);
        }
      });
      return;
    } catch (e) {
      if ((0, import_http_client2.isHttpClientAbortError)(e)) {
        throw e;
      }
      if (!(0, import_http_client2.isHttpClientResponseError)(e)) {
        throw new import_shared_errors4.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      if (response.status >= 400 && response.status <= 499) {
        const responseValue = await response.body("json");
        try {
          const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
          throw new ErrorEnvelopeError(`${response.status}`, response, errorBody, getErrorMessage(errorBody, ""), e);
        } catch (err) {
          if (err instanceof ErrorEnvelopeError) {
            throw err;
          }
          throw new import_shared_errors4.InvalidResponseError(
            `ResourceNotificationsClient.deleteResourceNotification:${response.status}`,
            err,
            responseValue,
            "ErrorEnvelope",
            void 0,
            response
          );
        }
      } else if (response.status >= 500 && response.status <= 599) {
        const responseValue = await response.body("json");
        try {
          const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
          throw new ErrorEnvelopeError(`${response.status}`, response, errorBody, getErrorMessage(errorBody, ""), e);
        } catch (err) {
          if (err instanceof ErrorEnvelopeError) {
            throw err;
          }
          throw new import_shared_errors4.InvalidResponseError(
            `ResourceNotificationsClient.deleteResourceNotification:${response.status}`,
            err,
            responseValue,
            "ErrorEnvelope",
            void 0,
            response
          );
        }
      } else {
        try {
          const responseValue = await response.body("json");
          throw new import_shared_errors4.ClientRequestError(
            `${response.status}`,
            response,
            responseValue,
            getErrorMessage(responseValue, `Unexpected api response: code=${response.status} body="${responseValue}"`)
          );
        } catch (error) {
          if (error instanceof import_shared_errors4.ClientRequestError) {
            throw error;
          }
          const responseValue = await response.body("text").catch(() => "");
          throw new import_shared_errors4.InvalidResponseError(
            `${response.status}`,
            error,
            responseValue,
            "json",
            "Unexpected api response",
            response
          );
        }
      }
    }
  }
};
var resourceNotificationsClient = /* @__PURE__ */ new ResourceNotificationsClient(import_http_client2.httpClient);

// packages/client/notification-v2/src/lib/error-envelopes/index.ts
var import_shared_errors5 = require("@dynatrace-sdk/shared-errors");

// packages/client/notification-v2/src/lib/models/davis-event-name-match.ts
var DavisEventNameMatch = /* @__PURE__ */ ((DavisEventNameMatch2) => {
  DavisEventNameMatch2["Equals"] = "equals";
  DavisEventNameMatch2["Contains"] = "contains";
  return DavisEventNameMatch2;
})(DavisEventNameMatch || {});

// packages/client/notification-v2/src/lib/models/davis-event-trigger-config-type.ts
var DavisEventTriggerConfigType = /* @__PURE__ */ ((DavisEventTriggerConfigType2) => {
  DavisEventTriggerConfigType2["DavisEvent"] = "davis-event";
  return DavisEventTriggerConfigType2;
})(DavisEventTriggerConfigType || {});

// packages/client/notification-v2/src/lib/models/davis-problem-trigger-config-type.ts
var DavisProblemTriggerConfigType = /* @__PURE__ */ ((DavisProblemTriggerConfigType2) => {
  DavisProblemTriggerConfigType2["DavisProblem"] = "davis-problem";
  return DavisProblemTriggerConfigType2;
})(DavisProblemTriggerConfigType || {});

// packages/client/notification-v2/src/lib/models/entity-tags-match.ts
var EntityTagsMatch = /* @__PURE__ */ ((EntityTagsMatch2) => {
  EntityTagsMatch2["All"] = "all";
  EntityTagsMatch2["Any"] = "any";
  return EntityTagsMatch2;
})(EntityTagsMatch || {});

// packages/client/notification-v2/src/lib/models/event-query-trigger-config-type.ts
var EventQueryTriggerConfigType = /* @__PURE__ */ ((EventQueryTriggerConfigType2) => {
  EventQueryTriggerConfigType2["Event"] = "event";
  return EventQueryTriggerConfigType2;
})(EventQueryTriggerConfigType || {});

// packages/client/notification-v2/src/lib/models/event-type.ts
var EventType = /* @__PURE__ */ ((EventType2) => {
  EventType2["Events"] = "events";
  EventType2["Bizevents"] = "bizevents";
  EventType2["DtSystemEvents"] = "dt.system.events";
  EventType2["SecurityEvents"] = "security.events";
  return EventType2;
})(EventType || {});

// packages/client/notification-v2/src/lib/models/maintenance-window-trigger-behavior-type.ts
var MaintenanceWindowTriggerBehaviorType = /* @__PURE__ */ ((MaintenanceWindowTriggerBehaviorType2) => {
  MaintenanceWindowTriggerBehaviorType2["Always"] = "always";
  MaintenanceWindowTriggerBehaviorType2["Inside"] = "inside";
  MaintenanceWindowTriggerBehaviorType2["Outside"] = "outside";
  return MaintenanceWindowTriggerBehaviorType2;
})(MaintenanceWindowTriggerBehaviorType || {});
