// @dynatrace/strato-components/content/markdown/transformers/table-mapping.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var table_mapping_exports = {};
__export(table_mapping_exports, {
  tableMappings: () => tableMappings
});
module.exports = __toCommonJS(table_mapping_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_spacings = __toESM(require("@dynatrace/strato-design-tokens/spacings"));
var import_SimpleTable = require("../../../tables/SimpleTable/SimpleTable.js");
var import_utils = require("../utils.js");
function getTextAlignment(cell) {
  return (0, import_react.isValidElement)(cell) ? cell?.props?.style?.textAlign : void 0;
}
const tableMappings = {
  table: ({ children, ...props }) => {
    if (children.length === 0) {
      return null;
    }
    const tHead = children[0];
    const tHeadRow = (0, import_utils.getChildren)(tHead)[0];
    const columns = (0, import_utils.getChildren)(
      tHeadRow
    ).map((cell, index) => ({
      header: () => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, { children: (0, import_utils.getChildren)(cell) }),
      // eslint-disable-next-line react/jsx-no-useless-fragment
      cell: ({ value }) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, { children: value }),
      accessor: String(index),
      id: String(index),
      alignment: getTextAlignment(cell)
    }));
    const tBody = children[1];
    const data = (0, import_utils.getChildren)(tBody).map((row) => ({
      ...(0, import_utils.getChildren)(row).map((cell) => (0, import_utils.getChildren)(cell))
    }));
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { ...props, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_SimpleTable.SimpleTable,
      {
        style: {
          marginTop: import_spacings.default.Size24,
          marginBottom: import_spacings.default.Size24
        },
        columns,
        data,
        variant: {
          contained: true,
          verticalDividers: true,
          rowDensity: "default"
        }
      }
    ) });
  }
};
