// @dynatrace/strato-components/content/markdown/transformers/typography-mapping.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var typography_mapping_exports = {};
__export(typography_mapping_exports, {
  typographyMappings: () => typographyMappings
});
module.exports = __toCommonJS(typography_mapping_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_typography_mapping_css = require("./typography-mapping.sty.js");
var import_Divider = require("../../../layouts/divider/Divider.js");
var import_Blockquote = require("../../../typography/block-quote/Blockquote.js");
var import_Code = require("../../../typography/code/Code.js");
var import_Emphasis = require("../../../typography/emphasis/Emphasis.js");
var import_ExternalLink = require("../../../typography/external-link/ExternalLink.js");
var import_Heading = require("../../../typography/heading/Heading.js");
var import_List = require("../../../typography/list/List.js");
var import_Paragraph = require("../../../typography/paragraph/Paragraph.js");
var import_Strikethrough = require("../../../typography/strikethrough/Strikethrough.js");
var import_Strong = require("../../../typography/strong/Strong.js");
var import_Text = require("../../../typography/text/Text.js");
var import_CodeSnippet = require("../../code-snippet/CodeSnippet.js");
var import_utils = require("../utils.js");
const MarkdownCode = (props) => {
  const { className, inline, children } = props;
  const match = (0, import_utils.matchLanguageParm)(className);
  let sanitizedChildren = children;
  if (Array.isArray(children)) {
    const firstChild = children[0];
    sanitizedChildren = [
      typeof firstChild === "string" ? firstChild.replace(/\n$/g, "") : ""
    ];
  } else if (typeof children === "string") {
    sanitizedChildren = children.replace(/\n$/g, "");
  }
  return inline ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Code.Code, { "data-testid": "code", children: sanitizedChildren }) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_CodeSnippet.CodeSnippet,
    {
      "data-testid": "code-snippet",
      language: match ? match[1] : void 0,
      children: sanitizedChildren
    }
  );
};
const typographyMappings = {
  code: MarkdownCode,
  // We need to add the inlineCode here since we use react-markdown in this repo
  // but mdx-js/react inside the cli-plugin
  inlineCode: ({ node, ...props }) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Code.Code, { ...props }),
  p: ({ children }) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Paragraph.Paragraph, { className: import_typography_mapping_css.MarkdownParagraphCSS, "data-testid": "paragraph", children }),
  strong: ({ ref, node, ...props }) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Strong.Strong, { "data-testid": "strong", ...props }),
  a: ({ ref, href, node, ...props }) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ExternalLink.ExternalLink,
    {
      "data-testid": "external-link",
      href,
      ...props
    }
  ),
  input: ({ checked }) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "span",
    {
      className: import_typography_mapping_css.MarkdownCheckmarkBackgroundCSS,
      role: "checkbox",
      "aria-label": checked ? "checked" : "unchecked",
      "aria-checked": !!checked,
      "aria-disabled": true,
      children: checked ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.CheckmarkSmallIcon, {}) : null
    }
  ),
  ul: ({ children }) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_List.List, { className: import_typography_mapping_css.MarkdownListCSS, children }),
  ol: ({ children }) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_List.List, { className: import_typography_mapping_css.MarkdownListCSS, ordered: true, children }),
  li: ({ ref, node, ordered, ...props }) => {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { ...props, className: import_typography_mapping_css.MarkdownListItemCSS });
  },
  h1: ({ ref, node, ...props }) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Heading.Heading,
    {
      ...props,
      className: import_typography_mapping_css.MarkdownHeadingCSS,
      "data-testid": "heading",
      level: 1,
      as: "h1"
    }
  ),
  h2: ({ ref, node, ...props }) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Heading.Heading,
    {
      ...props,
      className: import_typography_mapping_css.MarkdownHeadingCSS,
      "data-testid": "heading",
      level: 2,
      as: "h2"
    }
  ),
  h3: ({ ref, node, ...props }) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Heading.Heading,
    {
      ...props,
      className: import_typography_mapping_css.MarkdownHeadingCSS,
      "data-testid": "heading",
      level: 3,
      as: "h3"
    }
  ),
  h4: ({ ref, node, ...props }) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Heading.Heading,
    {
      ...props,
      className: import_typography_mapping_css.MarkdownHeadingCSS,
      "data-testid": "heading",
      level: 4,
      as: "h4"
    }
  ),
  h5: ({ ref, node, ...props }) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Heading.Heading,
    {
      ...props,
      className: import_typography_mapping_css.MarkdownHeadingCSS,
      "data-testid": "heading",
      level: 5,
      as: "h5"
    }
  ),
  h6: ({ ref, node, ...props }) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Heading.Heading,
    {
      ...props,
      className: import_typography_mapping_css.MarkdownHeadingCSS,
      "data-testid": "heading",
      level: 6,
      as: "h6"
    }
  ),
  del: ({ ref, node, ...props }) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Strikethrough.Strikethrough, { ...props }),
  em: ({ ref, node, ...props }) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Emphasis.Emphasis, { ...props }),
  blockquote: ({ ref, node, ...props }) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Blockquote.Blockquote, { ...props }),
  hr: () => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Divider.Divider, { orientation: "horizontal", as: "hr" }),
  img: ({ ref, alt, node, ...props }) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)("img", { ...props, alt, className: import_typography_mapping_css.MarkdownImageCSS })
};
