// @dynatrace/strato-components/content/markdown/transformers/typography-mapping.sty.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var typography_mapping_css_exports = {};
__export(typography_mapping_css_exports, {
  MarkdownCheckmarkBackgroundCSS: () => MarkdownCheckmarkBackgroundCSS,
  MarkdownHeadingCSS: () => MarkdownHeadingCSS,
  MarkdownImageCSS: () => MarkdownImageCSS,
  MarkdownListCSS: () => MarkdownListCSS,
  MarkdownListItemCSS: () => MarkdownListItemCSS,
  MarkdownParagraphCSS: () => MarkdownParagraphCSS
});
module.exports = __toCommonJS(typography_mapping_css_exports);
var import_typography_mapping_css_ts_vanilla = require("./typography-mapping.css");
var MarkdownCheckmarkBackgroundCSS = "_11vdepp5-3-6-0";
var MarkdownHeadingCSS = "_11vdepp0-3-6-0";
var MarkdownImageCSS = "_11vdepp4-3-6-0";
var MarkdownListCSS = "_11vdepp1-3-6-0";
var MarkdownListItemCSS = "_11vdepp2-3-6-0";
var MarkdownParagraphCSS = "_11vdepp3-3-6-0";
