// @dynatrace/strato-components/esm/forms/checkbox/Checkbox.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import {
  forwardRef,
  memo,
  useEffect,
  useImperativeHandle,
  useRef
} from "react";
import { CheckmarkSmallIcon, MinusIcon } from "@dynatrace/strato-icons";
import {
  checkboxIconCSS,
  checkboxKnobCSS,
  checkboxLabelCSS
} from "./Checkbox.sty.js";
import { VisuallyHidden } from "../../core/components/visually-hidden/VisuallyHidden.js";
import { useAriaLabelingProps as _useAriaLabelingProps } from "../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useControllableState } from "../../core/hooks/useControllableState.js";
import { useFormControlState } from "../../core/hooks/useFormControlState.js";
import { useSafeForwardProps as _useSafeForwardProps } from "../../core/hooks/useSafeForwardProps.js";
import { focusRingStatic } from "../../core/styles/focusRingStatic.sty.js";
import { mergeProps } from "../../core/utils/merge-props.js";
import { Flex } from "../../layouts/flex/Flex.js";
import { fieldTokens as fieldColorsCSS } from "../../styles/field.sty.js";
import { textStyle as textStyleCSS } from "../../styles/textStyle.sty.js";
import { InternalBaseItem as BaseItem } from "../forms-core/components/form-field-messages/FormFieldMessageBaseItem.js";
import { useFormFieldMessagesContext } from "../forms-core/hooks/useFormFieldMessagesContext.js";
import { useFormFieldValidation } from "../forms-core/hooks/useFormFieldValidation.js";
import { useSyncedFormControlProps } from "../forms-core/hooks/useSyncedFormControlProps.js";
import { FormFieldMessagesAction } from "../forms-core/state/formfield-messages-state-reducer.js";
import { useInputHint } from "../hooks/useInputHint.js";
const COMPONENT_NAME = "Checkbox";
const Checkbox = forwardRef(
  (props, forwardedRef) => {
    const {
      disabled: propDisabled,
      required: propRequired,
      readOnly: propReadOnly,
      id: propId,
      defaultValue,
      value,
      name,
      formValue,
      onChange,
      children,
      controlState,
      className: consumerClassName,
      style: consumerStyle,
      "aria-disabled": ariaDisabledProp,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
    const ariaLabelingProps = _useAriaLabelingProps(restProps);
    const forwardProps = _useSafeForwardProps(restProps, ariaLabelingProps);
    const { state: messagesContextState, dispatch: messagesContextDispatch } = useFormFieldMessagesContext();
    const {
      disabled,
      required,
      id,
      ariaDisabled,
      readOnly: _readOnly
    } = useSyncedFormControlProps({
      propDisabled,
      propRequired,
      propId,
      propAriaDisabled: ariaDisabledProp,
      propReadOnly
    });
    const rootRef = useRef(null);
    const inputRef = useRef(null);
    const {
      onChange: validationOnChange,
      onBlur: validationOnBlur,
      onInvalid: validationOnInvalid,
      ariaProps
    } = useFormFieldValidation(inputRef, { required }, ariaLabelingProps);
    useImperativeHandle(forwardedRef, () => {
      return {
        element: rootRef.current,
        inputRef: inputRef.current,
        focus() {
          inputRef.current?.focus();
        },
        validate() {
          if (messagesContextState) {
            inputRef.current?.checkValidity();
            messagesContextDispatch?.({
              type: FormFieldMessagesAction.SETISTOUCHED,
              payload: { isTouched: true }
            });
          }
        }
      };
    });
    const [state, setState] = useControllableState({
      defaultValue,
      value
    });
    const disabledOrAriaDisabled = disabled || ariaDisabled || false;
    const readOnly = !!_readOnly;
    const [hasError, hint] = useFormControlState(controlState);
    const { inputProps, hintProps } = useInputHint({
      hint,
      hasError
    });
    const showError = hasError || messagesContextState.showError;
    let focusRingColor = "neutral";
    if (showError) {
      focusRingColor = "critical";
    } else if (!(disabled || readOnly) && state) {
      focusRingColor = "primary";
    }
    const handleOnChange = (e) => {
      if (readOnly) {
        return;
      }
      const newValue = e.target.indeterminate ? true : e.target.checked;
      setState(newValue);
      onChange?.(newValue, e);
      validationOnChange();
      if (!messagesContextState.isTouched) {
        messagesContextDispatch?.({
          type: FormFieldMessagesAction.SETISTOUCHED,
          payload: { isTouched: true }
        });
      }
    };
    useEffect(() => {
      if (inputRef.current) {
        inputRef.current.indeterminate = value === "indeterminate";
      }
    }, [value]);
    let fieldColorVariant = "default";
    if (state === true || state === "indeterminate") {
      fieldColorVariant = "accent";
    } else if (showError || disabledOrAriaDisabled) {
      fieldColorVariant = "emphasized";
    }
    let fieldColorColor = "primary";
    if (showError) {
      fieldColorColor = "critical";
    } else if (state === false) {
      fieldColorColor = "neutral";
    }
    return /* @__PURE__ */ jsxs(
      Flex,
      {
        ref: rootRef,
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        flexDirection: "column",
        gap: 4,
        className: consumerClassName,
        style: consumerStyle,
        ...mergeProps(forwardProps, behavioralTrackingProps),
        "data-dt-component": COMPONENT_NAME,
        children: [
          /* @__PURE__ */ jsxs(
            "label",
            {
              htmlFor: id,
              className: clsx(
                checkboxLabelCSS({
                  required: typeof children === "string" && required,
                  disabled: disabledOrAriaDisabled,
                  readOnly
                }),
                textStyleCSS({ fontStyle: "text", textStyle: "base" })
              ),
              children: [
                /* @__PURE__ */ jsxs(
                  "div",
                  {
                    className: clsx(
                      checkboxKnobCSS({
                        disabled: disabledOrAriaDisabled,
                        hasChildren: !!children,
                        hasError: showError,
                        checked: state === true || state === "indeterminate",
                        readOnly
                      }),
                      fieldColorsCSS({
                        variant: fieldColorVariant,
                        disabled: disabledOrAriaDisabled,
                        color: fieldColorColor
                      }),
                      focusRingStatic({
                        when: "focus-within-visible",
                        color: focusRingColor,
                        disabled
                      })
                    ),
                    children: [
                      /* @__PURE__ */ jsx(VisuallyHidden, { children: /* @__PURE__ */ jsx(
                        "input",
                        {
                          id,
                          ref: inputRef,
                          type: "checkbox",
                          "aria-checked": state === "indeterminate" ? "mixed" : state,
                          "aria-required": required,
                          "aria-disabled": ariaDisabled,
                          readOnly,
                          name,
                          value: formValue,
                          disabled,
                          required,
                          checked: state === true,
                          onChange: handleOnChange,
                          onBlur: (event) => {
                            validationOnBlur(event);
                          },
                          onInvalid: validationOnInvalid,
                          ...controlState ? inputProps : ariaProps
                        }
                      ) }),
                      /* @__PURE__ */ jsx("div", { className: checkboxIconCSS({ readOnly, hasError: showError }), children: /* @__PURE__ */ jsx(CheckboxIconSvg, { state }) })
                    ]
                  }
                ),
                children
              ]
            }
          ),
          hint && /* @__PURE__ */ jsx(
            BaseItem,
            {
              variant: showError ? "error" : "hint",
              disabled: disabledOrAriaDisabled,
              ...hintProps,
              children: hint
            }
          )
        ]
      }
    );
  }
);
Checkbox.displayName = COMPONENT_NAME;
const CheckboxIconSvg = memo((props) => {
  switch (props.state) {
    case true:
      return /* @__PURE__ */ jsx(CheckmarkSmallIcon, {});
    case "indeterminate":
      return /* @__PURE__ */ jsx(MinusIcon, {});
    case false:
    default:
      return null;
  }
});
export {
  Checkbox
};
