// @dynatrace/strato-components/esm/charts/single-value/SingleValue.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import {
  forwardRef,
  useImperativeHandle,
  useRef
} from "react";
import { SINGLE_VALUE_ARIA_LABEL } from "./defaults.js";
import { useSingleValueConfig } from "./hooks/use-single-value-config.js";
import { container } from "./SingleValue.sty.js";
import { Sparkline } from "./slots/Sparkline.js";
import { Toolbar } from "./slots/Toolbar.js";
import { Trend } from "./slots/Trend.js";
import { iterateSlots } from "./utils/iterate-slots.js";
import { normalizeConfig } from "./utils/single-value-config.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useMergeRefs } from "../../core/hooks/useMergeRefs.js";
import { Intent } from "../../core/slots/Intent/Intent.js";
import { getChartStatesTemplates } from "../core/components/chart-states/utils/get-chart-states-template.js";
import { SingleValueRenderer } from "../core/components/single-value-renderer/SingleValueRenderer.js";
import { EmptyStateSlot } from "../core/slots/chart-states/EmptyStateSlot.js";
import { ErrorStateSlot } from "../core/slots/chart-states/ErrorStateSlot.js";
import { ColorRule } from "../core/slots/ColorRule.js";
import { ChartType } from "../core/types/chart-types.js";
import { coerceSizeValue } from "../core/utils/coerceSizeValue.js";
const COMPONENT_NAME = "SingleValue";
const SingleValueRoot = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const {
    data,
    prefixIcon,
    children,
    loading = false,
    width,
    height,
    style: customerStyle,
    className: customerClassNames,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "data-testid": dataTestId,
    "aria-label": ariaLabel,
    "aria-describedby": ariaDescribedBy,
    applyThresholdBackground,
    ...restProps
  } = props;
  const [behavioralTrackingProps] = useBehavioralTrackingProps(props);
  const { slots, sparklineChildren } = iterateSlots(children);
  const { emptyState, errorState } = getChartStatesTemplates(children);
  const config = useSingleValueConfig(restProps, slots);
  const sparklineProps = slots.sparkline ? {
    data: slots.sparkline?.data ?? [],
    ...config.sparkline
  } : void 0;
  const singleValueRef = useRef(null);
  const ref = useMergeRefs([
    forwardedRef,
    singleValueRef
  ]);
  useImperativeHandle(forwardedRef, () => ({
    getConfig: () => {
      return JSON.stringify({
        type: ChartType.SingleValue,
        ...normalizeConfig(config)
      });
    }
  }));
  const chartWidth = coerceSizeValue(width);
  const chartHeight = coerceSizeValue(height);
  const mergedStyle = {
    ...customerStyle,
    ...chartWidth && { width: chartWidth },
    ...chartHeight && { height: chartHeight }
  };
  return /* @__PURE__ */ jsx(
    "div",
    {
      ...behavioralTrackingProps,
      role: "img",
      className: clsx(container, customerClassNames),
      style: mergedStyle,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "data-testid": dataTestId,
      "data-dt-component": COMPONENT_NAME,
      ref,
      "aria-label": ariaLabel ?? SINGLE_VALUE_ARIA_LABEL,
      "aria-describedby": ariaDescribedBy,
      children: /* @__PURE__ */ jsx(
        SingleValueRenderer,
        {
          data,
          prefixIcon,
          ...config,
          sparklineProps,
          sparklineChildren,
          trendProps: slots.trend,
          emptyState,
          errorState,
          loading,
          applyThresholdBackground,
          children
        }
      )
    }
  );
});
SingleValueRoot.displayName = "SingleValue";
const SingleValue = Object.assign(SingleValueRoot, {
  Sparkline,
  Trend,
  ColorRule,
  EmptyState: EmptyStateSlot,
  ErrorState: ErrorStateSlot,
  Intent,
  Toolbar
});
export {
  SingleValue
};
