// @dynatrace/strato-components/forms/spin-buttons/number-spin-button/hooks/useBeforeInputHandler.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useBeforeInputHandler_exports = {};
__export(useBeforeInputHandler_exports, {
  useBeforeInputHandler: () => useBeforeInputHandler
});
module.exports = __toCommonJS(useBeforeInputHandler_exports);
var import_react = require("react");
var import_useValidationAndFormatting = require("./useValidationAndFormatting.js");
var import_useSelection = require("../../hooks/useSelection.js");
const reNumber = new RegExp("^-?\\d*$");
const reNonNegativeNumber = new RegExp("^\\d*$");
function getNrOfDigits(value) {
  return value < 0 ? 2 + Math.floor(Math.log10(Math.abs(value))) : 1 + Math.floor(Math.log10(value));
}
function processInputEvent(value, inputType, selection, subsitute) {
  const collapsed = selection.start === selection.end;
  switch (inputType) {
    case "insertText":
    case "insertFromPaste":
      return value.substring(0, selection.start) + subsitute + value.substring(selection.end);
    case "deleteByCut":
      return value.substring(0, selection.start) + value.substring(selection.end);
    case "deleteContentBackward":
      return collapsed ? value.substring(0, selection.start - 1) + value.substring(selection.end) : value.substring(0, selection.start) + value.substring(selection.end);
    case "deleteContentForward":
      return collapsed ? value.substring(0, selection.start) + value.substring(selection.end + 1) : value.substring(0, selection.start) + value.substring(selection.end);
    case "insertCompositionText":
      return value;
    case "deleteByDrag":
    case "insertFromDrop":
    default:
      return void 0;
  }
}
function getCandidate(text, domText, selection, inputType, substitute, formatLength) {
  const candidateOffset = formatLength > 0 ? Math.max(formatLength - text.length, 0) : 0;
  if (selection.start < candidateOffset && inputType === "deleteContentForward") {
    return void 0;
  }
  const useDomText = selection.start < candidateOffset && (inputType === "insertText" || inputType === "insertFromPaste");
  const value = useDomText ? domText.substring(selection.end) : text;
  const candidateSelection = {
    start: Math.max(selection.start - candidateOffset, 0),
    end: Math.max(selection.end - candidateOffset, 0)
  };
  return processInputEvent(value, inputType, candidateSelection, substitute);
}
function getNewSelection(selection, candidate, substitute, inputType, formatLength) {
  const rightAlign = formatLength > 0 && candidate.length <= formatLength;
  const collapsed = selection.start === selection.end;
  let curserPos = 0;
  switch (inputType) {
    case "insertText":
    case "insertFromPaste":
      curserPos = rightAlign ? selection.end : selection.start + substitute.length;
      break;
    case "deleteContentBackward":
      curserPos = rightAlign ? selection.end : collapsed ? selection.start - 1 : selection.start;
      break;
    case "deleteContentForward":
      curserPos = rightAlign ? collapsed ? selection.start + 1 : selection.end : selection.start;
      break;
    case "deleteByCut":
      curserPos = rightAlign ? selection.end : selection.start;
  }
  return {
    start: curserPos,
    end: curserPos
  };
}
const useBeforeInputHandler = (ref, props, applyText, text) => {
  const {
    onBeforeInput,
    displayFormat: { length: formatLength = 0 } = {},
    readOnly,
    disabled,
    onChange,
    minValue = Number.MIN_SAFE_INTEGER,
    maxValue = Number.MAX_SAFE_INTEGER
  } = props;
  const { getSelection, setSelection } = (0, import_useSelection.useSelection)(ref);
  const { parse, format } = (0, import_useValidationAndFormatting.useValidationAndFormatting)(
    props.displayFormat,
    props.required,
    minValue,
    maxValue
  );
  const defaultOnBeforeInput = (0, import_react.useCallback)(
    (candidate, event) => {
      const isInteger = minValue < 0 ? reNumber.test(candidate) : reNonNegativeNumber.test(candidate);
      if (!isInteger) {
        event.preventDefault();
        return;
      }
      const value = parse(candidate);
      if (value === null) {
        return;
      }
      const maxLength = value < 0 ? getNrOfDigits(minValue) : getNrOfDigits(maxValue);
      if (candidate.length > maxLength) {
        event.preventDefault();
        return;
      }
      if (value < 0 ? value < minValue : value > maxValue) {
        event.preventDefault();
        return;
      }
    },
    [maxValue, minValue, parse]
  );
  const beforeInputHandler = (0, import_react.useCallback)(
    (event) => {
      if (disabled || readOnly) {
        event.preventDefault();
        return;
      }
      const selection = getSelection();
      if (!selection) {
        event.preventDefault();
        return;
      }
      const domText = ref?.current?.textContent ?? "";
      const substitute = event.data ?? event.dataTransfer?.getData("text/plain") ?? "";
      const candidate = getCandidate(
        text,
        domText,
        selection,
        event.inputType,
        substitute,
        formatLength
      );
      if (candidate === void 0) {
        event.preventDefault();
        return;
      }
      onBeforeInput?.(candidate, event);
      if (event.defaultPrevented) {
        return;
      }
      defaultOnBeforeInput(candidate, event);
      if (event.defaultPrevented) {
        return;
      }
      const newSelection = domText === "" || domText === "-" ? { start: 0, end: 0, rtl: true } : getNewSelection(
        selection,
        candidate,
        substitute,
        event.inputType,
        formatLength
      );
      applyText(candidate);
      setSelection(newSelection);
      if (event.inputType === "insertCompositionText") {
        setTimeout(() => {
          if (ref?.current) {
            ref.current.textContent = format(text);
            setTimeout(() => setSelection(selection), 0);
          }
        }, 0);
      }
      const newValue = parse(candidate);
      props.value !== newValue && onChange?.(newValue);
      event.preventDefault();
    },
    [
      disabled,
      readOnly,
      getSelection,
      ref,
      text,
      formatLength,
      onBeforeInput,
      defaultOnBeforeInput,
      applyText,
      setSelection,
      parse,
      props.value,
      onChange,
      format
    ]
  );
  return beforeInputHandler;
};
