// @dynatrace/strato-components/charts/timeseries/components/displays/shared/utils/datapoint-stacking.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var datapoint_stacking_exports = {};
__export(datapoint_stacking_exports, {
  getCumulativeValue: () => getCumulativeValue,
  getStackedKeyByUnit: () => getStackedKeyByUnit
});
module.exports = __toCommonJS(datapoint_stacking_exports);
var import_units = require("@dynatrace-sdk/units");
var import_formatter = require("../../../../../core/types/formatter.js");
function getUnitString(unit) {
  if (unit === void 0) {
    return "undefined";
  }
  if (unit === null) {
    return "null";
  }
  if (unit === "") {
    return "";
  }
  if ((0, import_formatter.isUnit)(unit)) {
    return (0, import_units.formatUnit)(unit);
  }
  return unit.toString();
}
function getStackedKeyByUnit(sign, datapoint, unit) {
  return Math.sign(sign).toString() + datapoint.start.getTime() + datapoint.end?.getTime() + getUnitString(unit);
}
function getCumulativeValue(mapStackedPoints, datapoint, unit, seriesId, value = 0, sign) {
  const key = getStackedKeyByUnit(sign, datapoint, unit);
  const stackedDataPoint = mapStackedPoints.get(key);
  let datapointsCumulative = 0;
  let baseCumulative = 0;
  let cumulativeValue = value;
  if (stackedDataPoint) {
    const {
      datapointsCumulative: prevDatapointsCumulative,
      baseCumulative: prevBaseCumulative,
      seriesId: prevSeriesId
    } = stackedDataPoint;
    datapointsCumulative = prevDatapointsCumulative ?? 0;
    baseCumulative = prevBaseCumulative ?? 0;
    const baseCumulativeHasToChange = prevSeriesId !== seriesId;
    if (baseCumulativeHasToChange) {
      baseCumulative = datapointsCumulative;
    }
    cumulativeValue = baseCumulative + value;
  }
  datapointsCumulative = stackedDataPoint?.seriesId !== seriesId ? cumulativeValue : Math.sign(sign) > 0 ? Math.max(cumulativeValue, datapointsCumulative) : Math.min(cumulativeValue, datapointsCumulative);
  return { cumulativeValue, datapointsCumulative, baseCumulative, key };
}
