// @dynatrace/strato-components/esm/charts/gantt-chart/GanttChartLine.js
import { jsx, jsxs } from "react/jsx-runtime";
import { scaleLinear } from "d3-scale";
import {
  forwardRef,
  useState
} from "react";
import { useIntl } from "react-intl";
import { GanttChartLineRenderer } from "./components/GanttChartLineRenderer.js";
import { GanttChartTooltip } from "./components/tooltip/GanttChartTooltip.js";
import {
  DEFAULT_GANTT_LINE_SIZE,
  GANTT_CHART_V2_ARIA_LABEL,
  GANTT_LINE_SIZE_HEIGHT
} from "./constants/defaults.js";
import { GanttSegmentsContext } from "./context/GanttSegmentsContext.js";
import * as styles from "./GanttChartLine.sty.js";
import {
  useTooltipPinnedStatus,
  useTooltipVisibility
} from "./hooks/use-set-tooltip-state.js";
import { ganttChartMessages } from "./messages.js";
import { useMergeRefs } from "../../core/hooks/useMergeRefs.js";
import { useObserveContainerSize } from "../core/hooks/useObserveContainerSize.js";
const GanttChartLine = forwardRef((props, forwardedRef) => {
  const intl = useIntl();
  const [mouseInside, setMouseInside] = useState(false);
  const isTooltipVisible = useTooltipVisibility();
  const isTooltipPinned = useTooltipPinnedStatus();
  const { size, segmentsData, parentSegmentsData, config, loading } = props;
  const [{ width }, containerRef] = useObserveContainerSize();
  const ref = useMergeRefs([forwardedRef, containerRef]);
  const height = GANTT_LINE_SIZE_HEIGHT[size ?? DEFAULT_GANTT_LINE_SIZE];
  const segmentScale = scaleLinear().domain([props.min, props.max]).range([0, width]);
  return /* @__PURE__ */ jsx(
    "div",
    {
      className: styles.wrapper,
      onMouseEnter: () => !isTooltipPinned && setMouseInside(true),
      onMouseLeave: () => !isTooltipPinned && setMouseInside(false),
      role: "img",
      "aria-label": GANTT_CHART_V2_ARIA_LABEL,
      children: /* @__PURE__ */ jsxs(
        GanttSegmentsContext.Provider,
        {
          value: {
            segmentsData,
            parentSegmentsData,
            segmentScale,
            config,
            hovering: mouseInside
          },
          children: [
            /* @__PURE__ */ jsxs(
              GanttChartLineRenderer,
              {
                width,
                height,
                loading,
                ref,
                showBackground: config.showBackground,
                children: [
                  /* @__PURE__ */ jsx(GanttChartLineRenderer.EmptyState, { children: intl.formatMessage(ganttChartMessages.emptyState) }),
                  /* @__PURE__ */ jsx(GanttChartLineRenderer.ErrorState, { children: (errorMessage) => /* @__PURE__ */ jsxs("span", { children: [
                    intl.formatMessage(ganttChartMessages.errorState),
                    ":",
                    " ",
                    errorMessage
                  ] }) })
                ]
              }
            ),
            isTooltipVisible && /* @__PURE__ */ jsx(GanttChartTooltip, {})
          ]
        }
      )
    }
  );
});
export {
  GanttChartLine
};
