// @dynatrace/strato-components/esm/forms/radio/RadioGroup.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import {
  forwardRef,
  useImperativeHandle,
  useMemo,
  useRef
} from "react";
import { RadioGroupContext } from "./RadioGroupContext.js";
import * as styles from "./styles.sty.js";
import { useControllableState } from "../../core/hooks/useControllableState.js";
import { useFormControlState } from "../../core/hooks/useFormControlState.js";
import { isDev } from "../../core/utils/is-node-environment.js";
import { Flex } from "../../layouts/flex/Flex.js";
import { withFormFieldMessageContextProvider } from "../forms-core/components/form-field/FormFieldMessageContextProvider.js";
import { InternalBaseItem as BaseItem } from "../forms-core/components/form-field-messages/FormFieldMessageBaseItem.js";
import { useFormFieldMessagesContext } from "../forms-core/hooks/useFormFieldMessagesContext.js";
import { useSyncedFormControlProps } from "../forms-core/hooks/useSyncedFormControlProps.js";
import { FormFieldMessagesAction } from "../forms-core/state/formfield-messages-state-reducer.js";
import { useInputHint } from "../hooks/useInputHint.js";
const RadioGroupRoot = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const {
    children,
    id: propId,
    disabled: propDisabled,
    value,
    name,
    defaultValue,
    controlState,
    onChange,
    readOnly: propReadOnly,
    required: propRequired,
    "aria-disabled": propAriaDisabled,
    "aria-describedby": ariaDescribedby,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    ...remainingProps
  } = props;
  const {
    disabled,
    required,
    id,
    ariaDisabled,
    readOnly: _readOnly
  } = useSyncedFormControlProps({
    propDisabled,
    propRequired,
    propId,
    propAriaDisabled,
    propReadOnly
  });
  const readOnly = !!_readOnly;
  const { state: messagesContextState, dispatch: messagesContextDispatch } = useFormFieldMessagesContext();
  const [controlStateHasError, hint] = useFormControlState(controlState);
  const hasError = controlState ? controlStateHasError : messagesContextState.showError;
  const { inputProps, hintProps } = useInputHint({
    ariaDescribedby,
    hint,
    hasError
  });
  const [controllableState, setControllableState] = useControllableState({
    defaultValue,
    value,
    onChange
  });
  const wrapperRef = useRef(null);
  useImperativeHandle(forwardedRef, () => {
    return {
      element: wrapperRef.current,
      focus() {
        const checkedInput = wrapperRef.current?.querySelector(
          "input:checked"
        );
        (checkedInput ?? wrapperRef.current?.querySelector("input"))?.focus();
      },
      validate() {
        if (controlState !== void 0) {
          return;
        }
        const inputs = wrapperRef.current?.querySelectorAll("input");
        inputs?.forEach((input) => {
          input?.checkValidity();
        });
        messagesContextDispatch?.({
          type: FormFieldMessagesAction.SETISTOUCHED,
          payload: { isTouched: true }
        });
      }
    };
  });
  if (isDev() && name === void 0) {
    console.warn(
      'RadioGroup: The "name" prop is missing, which is necessary to link the radio buttons correctly.'
    );
  }
  const radioGroupContextValue = useMemo(
    () => ({
      name,
      isDisabled: disabled,
      ariaDisabled,
      hasError,
      value: controllableState,
      onChange: setControllableState,
      readOnly,
      required,
      controlState
    }),
    [
      name,
      disabled,
      ariaDisabled,
      hasError,
      controllableState,
      setControllableState,
      readOnly,
      required,
      controlState
    ]
  );
  return /* @__PURE__ */ jsxs(
    "div",
    {
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "aria-orientation": "vertical",
      role: "radiogroup",
      "aria-live": "polite",
      "aria-disabled": propAriaDisabled,
      ref: wrapperRef,
      className: clsx(consumerClassName, styles.group, {
        [styles.readOnly]: readOnly,
        [styles.hasError]: hasError
      }),
      style: consumerStyle,
      id,
      ...inputProps,
      ...remainingProps,
      children: [
        /* @__PURE__ */ jsx(RadioGroupContext.Provider, { value: radioGroupContextValue, children: /* @__PURE__ */ jsx(Flex, { flexDirection: "column", gap: 8, children }) }),
        hint && /* @__PURE__ */ jsx(
          BaseItem,
          {
            variant: hasError ? "error" : "hint",
            disabled: disabled || ariaDisabled,
            ...hintProps,
            children: hint
          }
        )
      ]
    }
  );
});
const RadioGroup = withFormFieldMessageContextProvider(RadioGroupRoot);
RadioGroup.displayName = "RadioGroup";
export {
  RadioGroup
};
