// @dynatrace/strato-components/esm/forms/radio/Radio.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import {
  forwardRef,
  useContext,
  useEffect,
  useImperativeHandle,
  useRef
} from "react";
import { RadioGroupContext } from "./RadioGroupContext.js";
import * as styles from "./styles.sty.js";
import { VisuallyHidden } from "../../core/components/visually-hidden/VisuallyHidden.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useId } from "../../core/hooks/useId.js";
import { useFocusRing } from "../../core/styles/useFocusRing.js";
import { isDev } from "../../core/utils/is-node-environment.js";
import { mergeProps as _mergeProps } from "../../core/utils/merge-props.js";
import { fieldTokens as fieldColorsCSS } from "../../styles/field.sty.js";
import { Text } from "../../typography/text/Text.js";
import { useFormFieldMessagesContext } from "../forms-core/hooks/useFormFieldMessagesContext.js";
import { useFormFieldValidation } from "../forms-core/hooks/useFormFieldValidation.js";
import { FormFieldMessagesAction } from "../forms-core/state/formfield-messages-state-reducer.js";
const COMPONENT_NAME = "Radio";
const Radio = /* @__PURE__ */ forwardRef(
  (props, forwardedRef) => {
    const fallbackId = useId();
    const {
      disabled: propDisabled,
      id = fallbackId,
      children,
      value,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "aria-label": ariaLabelProp,
      "aria-labelledby": ariaLabelledByProp,
      "aria-describedby": ariaDescribedByProp,
      "aria-details": ariaDetailsProp,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
    const { dispatch: messagesContextDispatch } = useFormFieldMessagesContext();
    const radioGroupContext = useContext(RadioGroupContext);
    if (isDev() && !radioGroupContext) {
      console.warn(
        "<Radio /> can only be used as a child element of <RadioGroup />"
      );
    }
    const {
      isDisabled,
      ariaDisabled,
      controlState,
      hasError,
      readOnly,
      name: groupName,
      required: groupRequired,
      value: groupValue,
      onChange
    } = radioGroupContext || {};
    const radioLabelRef = useRef(null);
    const inputRef = useRef(null);
    useImperativeHandle(forwardedRef, () => {
      return {
        element: radioLabelRef.current,
        inputRef: inputRef.current,
        focus() {
          inputRef.current?.focus();
        },
        validate() {
          if (controlState !== void 0) {
            return;
          }
          inputRef.current?.checkValidity();
          messagesContextDispatch?.({
            type: FormFieldMessagesAction.SETISTOUCHED,
            payload: { isTouched: true }
          });
        }
      };
    });
    const {
      onChange: validationOnChange,
      onBlur: validationOnBlur,
      onInvalid: validationOnInvalid,
      ariaProps
    } = useFormFieldValidation(
      inputRef,
      { groupRequired },
      {
        "aria-label": ariaLabelProp,
        "aria-labelledby": ariaLabelledByProp,
        "aria-describedby": ariaDescribedByProp,
        "aria-details": ariaDetailsProp
      }
    );
    useEffect(() => {
      messagesContextDispatch?.({
        type: FormFieldMessagesAction.ADDFORMCONTROLID,
        payload: { controlId: id }
      });
      return () => messagesContextDispatch?.({
        type: FormFieldMessagesAction.REMOVEFORMCONTROLID,
        payload: { controlId: id }
      });
    }, [id]);
    const disabled = propDisabled ?? isDisabled;
    const disabledOrAriaDisabled = disabled || ariaDisabled || false;
    const checked = value === groupValue;
    function handleChange(e) {
      if (readOnly) {
        return;
      }
      onChange?.(e.target.value);
      !controlState && validationOnChange?.();
    }
    let variant = "default";
    let color = "neutral";
    if (checked) {
      variant = "accent";
    } else if (hasError) {
      variant = "emphasized";
    }
    if (hasError) {
      color = "critical";
    } else if (!readOnly && checked) {
      color = "primary";
    }
    const { focusProps, focusClassName, isFocusVisible } = useFocusRing({
      color,
      disabled
    });
    return /* @__PURE__ */ jsxs(
      "label",
      {
        "data-testid": dataTestId,
        htmlFor: id,
        ref: radioLabelRef,
        className: clsx(
          styles.label,
          {
            [styles.disabled]: disabledOrAriaDisabled,
            [styles.checked]: checked
          },
          consumerClassName
        ),
        style: consumerStyle,
        ...behavioralTrackingProps,
        "data-dt-component": COMPONENT_NAME,
        children: [
          /* @__PURE__ */ jsx(VisuallyHidden, { children: /* @__PURE__ */ jsx(
            "input",
            {
              ref: inputRef,
              checked,
              name: groupName,
              value,
              type: "radio",
              disabled,
              "aria-disabled": ariaDisabled,
              required: groupRequired,
              id,
              onChange: handleChange,
              onInvalid: (e) => {
                if (controlState) {
                  e.preventDefault();
                  return;
                }
                validationOnInvalid(e);
              },
              readOnly,
              ..._mergeProps(
                restProps,
                focusProps,
                controlState ? {} : {
                  ...ariaProps,
                  onBlur: validationOnBlur
                }
              )
            }
          ) }),
          /* @__PURE__ */ jsx(
            "div",
            {
              className: clsx(
                styles.knob({
                  disabled: disabledOrAriaDisabled,
                  checked,
                  isFocusVisible,
                  hasError,
                  readOnly
                }),
                fieldColorsCSS({
                  disabled: disabledOrAriaDisabled,
                  variant,
                  color
                }),
                focusClassName
              ),
              children: /* @__PURE__ */ jsx(
                "svg",
                {
                  className: clsx(
                    styles.innerKnob,
                    fieldColorsCSS({
                      color: hasError ? "critical" : "primary",
                      disabled: disabledOrAriaDisabled,
                      variant: "accent"
                    })
                  ),
                  viewBox: "0 0 100 100",
                  children: /* @__PURE__ */ jsx("circle", { r: isFocusVisible ? "20" : "25", cx: "50", cy: "50" })
                }
              )
            }
          ),
          /* @__PURE__ */ jsx(Text, { as: "span", textStyle: "base", children })
        ]
      }
    );
  }
);
Radio.displayName = COMPONENT_NAME;
export {
  Radio
};
