// @dynatrace/strato-components/esm/forms/radio/styles.sty.js
import "../../core/components/app-root/AppRoot.css";
import "../../styles/field.css";
import "./styles.css";
import { createRuntimeFn as _7a468 } from "@vanilla-extract/recipes/createRuntimeFn";
var checked = "_h96idu4-3-6-0";
var disabled = "_h96idu3-3-6-0";
var group = "_h96idu0-3-6-0";
var hasError = "_h96idu2-3-6-0";
var innerKnob = "_h96iduk-3-6-0";
var knob = _7a468({ defaultClassName: "_h96idu6-3-6-0", variantClassNames: { disabled: { true: "_h96idu7-3-6-0" }, checked: { true: "_h96idu8-3-6-0", false: "_h96idu9-3-6-0" }, hasError: { true: "_h96idua-3-6-0", false: "_h96idub-3-6-0" }, isFocusVisible: { true: "_h96iduc-3-6-0", false: "_h96idud-3-6-0" }, readOnly: { true: "_h96idue-3-6-0", false: "_h96iduf-3-6-0" } }, defaultVariants: {}, compoundVariants: [[{ checked: false, hasError: true, isFocusVisible: false, disabled: false, readOnly: false }, "_h96idug-3-6-0"], [{ checked: false, hasError: false, isFocusVisible: false, disabled: false, readOnly: false }, "_h96iduh-3-6-0"], [{ readOnly: true, disabled: false }, "_h96idui-3-6-0"], [{ readOnly: true, hasError: true, disabled: false }, "_h96iduj-3-6-0"]] });
var label = "_h96idu5-3-6-0";
var readOnly = "_h96idu1-3-6-0";
export {
  checked,
  disabled,
  group,
  hasError,
  innerKnob,
  knob,
  label,
  readOnly
};
