// @dynatrace/strato-components/charts/single-value-grid/components/SingleValueDisplay.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SingleValueDisplay_exports = {};
__export(SingleValueDisplay_exports, {
  SingleValueDisplay: () => SingleValueDisplay
});
module.exports = __toCommonJS(SingleValueDisplay_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_react_intl = require("react-intl");
var import_SingleValueRenderer = require("../../core/components/single-value-renderer/SingleValueRenderer.js");
var import_messages = require("../messages.js");
var import_build_config = require("../utils/build-config.js");
var import_build_props_from_grid = require("../utils/build-props-from-grid.js");
var import_get_config_from_accessors = require("../utils/get-config-from-accessors.js");
var import_get_data_from_accessors = require("../utils/get-data-from-accessors.js");
const SingleValueDisplay = ({
  record,
  propsConfig,
  accessors
}) => {
  const intl = (0, import_react_intl.useIntl)();
  const gridIntents = propsConfig.intents.grid;
  const singleIntents = propsConfig.intents.single;
  const propsFromGrid = (0, import_build_props_from_grid.buildPropsFromGrid)(
    propsConfig.toolbar,
    singleIntents,
    gridIntents,
    intl.formatMessage(import_messages.SingleValueGridMessages.gridIntentsLabel)
  );
  if (typeof record === "string" || typeof record === "number") {
    const config2 = (0, import_build_config.buildConfig)(propsConfig);
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_SingleValueRenderer.SingleValueRenderer,
      {
        data: record,
        color: config2.color,
        formatter: config2.value?.formatter,
        unit: config2.value?.unit,
        colorRules: config2.colorRules,
        alignment: config2.alignment,
        trendProps: config2.trend,
        applyThresholdBackground: config2.applyThresholdBackground,
        propsFromGrid: propsFromGrid({ value: record })
      }
    );
  }
  const { label, prefixIcon, value, trendData, sparkline } = (0, import_get_data_from_accessors.getDataFromAccessors)(record, accessors);
  const accessorsConfig = (0, import_get_config_from_accessors.getConfigFromAccessors)(record, accessors);
  const config = (0, import_build_config.buildConfig)(propsConfig, accessorsConfig);
  let sparklineProps = void 0;
  if (!(0, import_lodash_es.isUndefined)(sparkline?.data)) {
    sparklineProps = {
      ...config.sparkline,
      data: sparkline.data,
      xAxis: sparkline.xAxis,
      yAxis: sparkline.yAxis
    };
  }
  let trendProps = void 0;
  if (!(0, import_lodash_es.isUndefined)(config.trend) || !(0, import_lodash_es.isUndefined)(trendData)) {
    trendProps = {
      ...config.trend,
      value: trendData?.value,
      label: trendData?.label,
      direction: trendData?.direction
    };
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_SingleValueRenderer.SingleValueRenderer,
    {
      data: value,
      dataItem: record,
      label,
      prefixIcon,
      color: config.color,
      formatter: config.value?.formatter,
      unit: config.value?.unit,
      colorRules: config.colorRules,
      alignment: config.alignment,
      sparklineProps,
      trendProps,
      applyThresholdBackground: config.applyThresholdBackground,
      propsFromGrid: propsFromGrid(record)
    }
  );
};
