// @dynatrace/strato-components/esm/tables/DataTable/features/ColumnVisibility/ColumnVisibility.js
import { useMemo } from "react";
import { useControllableState } from "../../../../core/hooks/useControllableState.js";
import { isDev } from "../../../../core/utils/is-node-environment.js";
import { useDataTableConfigElement } from "../../../hooks/useDataTableConfig.js";
import { isGroupColumnDef } from "../../hooks/useTable/datatable-typeguards.js";
import { getMemoOptions } from "../../hooks/useTable/utils.js";
import { memo } from "../../tanstack-table-core-override/src/index.js";
function getCanHide(table, column, isHideable) {
  if (table.options.enableHiding === false) {
    return false;
  }
  if (!column.columns || column.columns.length === 0) {
    return column.columnDef.enableHiding ?? true;
  }
  return column.columnDef.enableHiding === false ? false : isHideable(column.columns);
}
function isParentColumnHideable(childColumns) {
  return childColumns.every((c) => c.columnDef.enableHiding);
}
const DataTableColumnVisibility = {
  createTable(table) {
    table.getOrderedVisibleLeafColumns = memo(
      () => {
        const state = table.getState();
        return [
          state.columnVisibility,
          state.columnPinning,
          state.columnOrder,
          table.options.columns
        ];
      },
      () => {
        return [
          ...table.getLeftVisibleLeafColumns(),
          ...table.getCenterVisibleLeafColumns(),
          ...table.getRightVisibleLeafColumns()
        ];
      },
      { key: isDev() && "getOrderedVisibleLeafColumns" }
    );
    table.getVisibleLeafColumnsWithoutBuiltin = memo(
      () => {
        const state = table.getState();
        return [
          state.columnVisibility,
          state.columnOrder,
          table.options.columns
        ];
      },
      () => {
        return table.getVisibleLeafColumns().filter((leafColumn) => !leafColumn.columnDef.meta?.builtin);
      },
      {
        key: isDev() && "getVisibleLeafColumnsWithoutBuiltin"
      }
    );
    table.getCenterVisibleLeafColumnsWithoutBuiltIn = memo(
      () => {
        const state = table.getState();
        return [
          state.columnVisibility,
          state.columnPinning,
          state.columnOrder,
          table.options.columns
        ];
      },
      () => {
        return table.getCenterVisibleLeafColumns().filter((leafColumn) => !leafColumn.columnDef.meta?.builtin);
      },
      { key: isDev() && "getCenterVisibleLeafColumnsWithoutBuildIn" }
    );
    table.getVisibleVirtualColumnIdsWithoutBuiltin = () => {
      const visibleLeafColumns = table.getVisibleLeafColumnsWithoutBuiltin();
      const virtualColumns = table.getColumnVirtualizer()?.getVirtualItems() ?? [];
      return virtualColumns.flatMap(
        (virtualItem) => visibleLeafColumns.at(virtualItem.index)?.id ?? []
      );
    };
  },
  createColumn(column, table) {
    column.getIsLastVisibleColumn = memo(
      () => {
        const state = table.getState();
        return [
          state.columnVisibility,
          state.columnPinning,
          table.options.columns
        ];
      },
      () => {
        const lastVisibleLeafColumnId = table.getRightVisibleLeafColumns().at(-1)?.id ?? table.getCenterVisibleLeafColumns().at(-1)?.id;
        return column.id === lastVisibleLeafColumnId || column.columns.some(
          (childColumn) => childColumn.id === lastVisibleLeafColumnId
        );
      },
      getMemoOptions(
        table.options,
        "debugColumns",
        "column.getIsLastVisibleColumn"
      )
    );
    column.getVisibleLeafColumns = memo(
      () => {
        const state = table.getState();
        return [
          state.columnVisibility,
          state.columnOrder,
          table.options.columns
        ];
      },
      () => {
        return column.getLeafColumns().filter((leafColumn) => leafColumn.getIsVisible());
      },
      getMemoOptions(
        table.options,
        "debugColumns",
        "column.getVisibleLeafColumns"
      )
    );
    column.getCenterVisibleLeafColumns = memo(
      () => {
        const state = table.getState();
        return [
          state.columnVisibility,
          state.columnPinning,
          state.columnOrder,
          table.options.columns
        ];
      },
      () => {
        return column.getVisibleLeafColumns().filter((leafColumn) => !leafColumn.getIsPinned());
      },
      getMemoOptions(
        table.options,
        "debugColumns",
        "column.getCenterVisibleLeafColumns"
      )
    );
    column.getLeftVisibleLeafColumns = memo(
      () => {
        const state = table.getState();
        return [
          state.columnVisibility,
          state.columnPinning.left,
          state.columnOrder,
          table.options.columns
        ];
      },
      () => {
        const ids = new Set(column.getVisibleLeafColumns().map((c) => c.id));
        return table.getLeftVisibleLeafColumns().filter((c) => ids.has(c.id));
      },
      getMemoOptions(
        table.options,
        "debugColumns",
        "column.getLeftVisibleLeafColumns"
      )
    );
    column.getRightVisibleLeafColumns = memo(
      () => {
        const state = table.getState();
        return [
          state.columnVisibility,
          state.columnPinning.right,
          state.columnOrder,
          table.options.columns
        ];
      },
      () => {
        const ids = new Set(column.getVisibleLeafColumns().map((c) => c.id));
        return table.getRightVisibleLeafColumns().filter((c) => ids.has(c.id));
      },
      getMemoOptions(
        table.options,
        "debugColumns",
        "column.getRightVisibleLeafColumns"
      )
    );
    column.getCanHide = () => getCanHide(table, column, isParentColumnHideable);
  }
};
function useColumnVisibility(props, options) {
  const columnVisibilityConfig = useDataTableConfigElement("columnVisibility");
  const [columnVisibility, onColumnVisibilityChange] = useControllableState({
    value: props.columnVisibility ?? columnVisibilityConfig,
    defaultValue: props.defaultColumnVisibility,
    onChange: props.onColumnVisibilityChange
  });
  options.state = { ...options.state, columnVisibility };
  options.onColumnVisibilityChange = onColumnVisibilityChange;
  options.originalColumnVisibility = useMemo(() => {
    return props.defaultColumnVisibility ?? getColumnVisibilityFromOriginalColumns(options.columns);
  }, [props.defaultColumnVisibility, options.columns]);
}
function getColumnVisibilityFromOriginalColumns(columnDefs) {
  const visibilityState = {};
  for (const columnDef of columnDefs) {
    if (isGroupColumnDef(columnDef) && columnDef.columns) {
      const subColumnDefs = columnDef.columns;
      for (const subColumnDef of subColumnDefs) {
        if (subColumnDef.id) {
          visibilityState[subColumnDef.id] = true;
        }
      }
    } else if (columnDef.id) {
      visibilityState[columnDef.id] = true;
    }
  }
  return visibilityState;
}
export {
  DataTableColumnVisibility,
  getCanHide,
  useColumnVisibility
};
