// @dynatrace/strato-components/editors/code-editor/auto-completion/javascript.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var javascript_exports = {};
__export(javascript_exports, {
  javascriptAutocompleteProvider: () => javascriptAutocompleteProvider
});
module.exports = __toCommonJS(javascript_exports);
var import_autocomplete = require("@codemirror/autocomplete");
var import_lang_javascript = require("@codemirror/lang-javascript");
const keywords = "break case const continue default delete export extends false finally in instanceof let new return static super switch this throw true typeof var yield".split(" ").map((kw) => ({
  label: kw,
  type: "keyword"
}));
const dontComplete = [
  "TemplateString",
  "String",
  "RegExp",
  "LineComment",
  "BlockComment",
  "VariableDefinition",
  "TypeDefinition",
  "Label",
  "PropertyDefinition",
  "PropertyName",
  "PrivatePropertyDefinition",
  "PrivatePropertyName"
];
const javascriptAutocompleteProvider = async (context) => {
  const localCompletionResult = (0, import_lang_javascript.localCompletionSource)(context);
  const javascriptCompletionResult = await javascriptCompletionSource()(
    context
  );
  const scopeCompletionResult = await (0, import_lang_javascript.scopeCompletionSource)(globalThis)(
    context
  );
  const before = context.matchBefore(/\w+/);
  const completions = extractCompletions([
    localCompletionResult,
    javascriptCompletionResult,
    scopeCompletionResult
  ]);
  if (completions.length === 0) {
    return { from: context.pos, suggestions: [] };
  }
  return {
    from: before ? before.from : context.pos,
    keepOrder: true,
    validFor: /\w+/,
    suggestions: transformSuggestions(completions)
  };
};
const javascriptCompletionSource = () => {
  return (0, import_autocomplete.ifNotIn)(dontComplete, (0, import_autocomplete.completeFromList)(import_lang_javascript.snippets.concat(keywords)));
};
const extractCompletions = (completionResultList) => {
  let completions = [];
  for (const completionResult of completionResultList) {
    if (completionResult !== null) {
      completions = completions.concat(completionResult.options);
    }
  }
  return completions;
};
const transformSuggestions = (completions) => {
  const suggestions = [];
  for (const completion of completions) {
    suggestions.push({
      postfix: completion.detail,
      label: completion.label,
      type: completion.type,
      apply: completion.apply
    });
  }
  return suggestions;
};
