// @dynatrace/strato-components/tables/DataTable/features/Download/Download.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Download_exports = {};
__export(Download_exports, {
  DataTableDownload: () => DataTableDownload,
  DataTableDownloadData: () => DataTableDownloadData,
  useDownloadData: () => useDownloadData
});
module.exports = __toCommonJS(Download_exports);
var import_download = require("../../../../core/utils/download.js");
var import_logging = require("../../../../core/utils/logging.js");
var import_download_utils = require("../../../utils/download-utils.js");
var import_constants = require("../../constants.js");
function DataTableDownload(props) {
  return null;
}
function getCsvHeaders(table, includeIndexColumnHeader, excludeColumns) {
  const columnHeaders = table.getLeafHeaders().filter((header) => {
    if (header.subHeaders.length > 0) {
      return false;
    }
    if (header.isPlaceholder) {
      return false;
    }
    if (header.column.columnDef.meta?.builtin === true) {
      return false;
    }
    if (excludeColumns?.includes(header.column.id)) {
      return false;
    }
    if (header.column.columnDef.columnType === "sparkline") {
      return false;
    }
    if (header.column.getIsVisible() === false) {
      return false;
    }
    return true;
  }).map((header) => header._getDownloadString());
  if (includeIndexColumnHeader) {
    columnHeaders.unshift(import_constants.DATA_TABLE_INDEX_COLUMN_HEADER);
  }
  return columnHeaders;
}
function getCsvCells(row, excludeColumns) {
  return row.getVisibleCells().filter(
    (cell) => cell.column.columnDef.meta?.builtin !== true && cell.column.columnDef.columnType !== "sparkline" && !excludeColumns?.includes(cell.column.id)
  ).map((cell) => cell._getDownloadString());
}
function getRowIndex(row) {
  let indexPrefix = "";
  for (const parent of row.getParentRows()) {
    indexPrefix = `${indexPrefix}${parent.index + 1}.`;
  }
  return `${indexPrefix}${row.index + 1}`;
}
async function getCsv(table, rowModel, excludeColumns) {
  const nrOfRows = rowModel.flatRows.length;
  const csvData = Array.from({ length: nrOfRows + 1 });
  const batchSize = 1e3;
  let currentIndex = 0;
  const includeIndexColumn = table.getHasSubRows();
  csvData[0] = getCsvHeaders(table, includeIndexColumn, excludeColumns);
  function processBatch() {
    return new Promise((resolve) => {
      const currentBatchEnd = Math.min(currentIndex + batchSize, nrOfRows);
      for (let i = currentIndex; i < currentBatchEnd; i++) {
        const row = rowModel.flatRows[i];
        const csvRow = getCsvCells(row, excludeColumns);
        if (includeIndexColumn) {
          const rowIndex = getRowIndex(row);
          csvRow.unshift(rowIndex);
        }
        csvData[i + 1] = csvRow;
      }
      currentIndex = currentBatchEnd;
      if (currentIndex < nrOfRows) {
        setTimeout(() => resolve(processBatch()), 0);
      } else {
        resolve();
      }
    });
  }
  await processBatch();
  return csvData;
}
function downloadData(table, rowModel, subset, excludeColumns, onDownloadEnd) {
  getCsv(table, rowModel, excludeColumns).then((rows) => {
    const csvString = (0, import_download_utils.convertArrayToCsv)(rows);
    (0, import_download.downloadCSV)(csvString, import_constants.DATA_TABLE_DOWNLOAD_CSV_FILE_NAME);
    table.options.onDownloadData?.(
      subset,
      ...excludeColumns ? [excludeColumns] : []
    );
  }).finally(() => {
    onDownloadEnd?.();
  });
}
const DataTableDownloadData = {
  createHeader(header, _table) {
    header._getDownloadString = () => {
      const nestedHeaderNames = [];
      let currentColumnNesting = header.column;
      while (currentColumnNesting !== void 0) {
        const columnDefHeader = currentColumnNesting.columnDef.header;
        if (typeof columnDefHeader === "function" || columnDefHeader === void 0) {
          nestedHeaderNames.push(header.id);
        } else {
          nestedHeaderNames.push(columnDefHeader);
        }
        currentColumnNesting = currentColumnNesting.parent;
      }
      return nestedHeaderNames.toReversed().join(".");
    };
  },
  createCell(cell, column, _row, _table) {
    cell._getDownloadString = () => {
      const cellValue = cell.getValue();
      if (cellValue === void 0 || cellValue === null) {
        return "";
      }
      if (cellValue instanceof Date) {
        return cellValue.toISOString();
      }
      if ((column.columnDef.columnType === "datetime" || // DEPRECATION: [APPDEV-16044] BREAKING-CHANGE: Remove 'date' columnType.
      column.columnDef.columnType === "date") && (typeof cellValue === "string" || typeof cellValue === "number")) {
        const cellDate = new Date(cellValue);
        if (!isNaN(cellDate.getTime())) {
          return cellDate.toISOString();
        }
        if (typeof cellValue === "string" && !isNaN(parseInt(cellValue, 10))) {
          const cellDate2 = new Date(parseInt(cellValue, 10));
          if (!isNaN(cellDate2.getTime())) {
            return cellDate2.toISOString();
          }
        }
      }
      if (column.columnDef.columnType === "meterbar") {
        if (!Array.isArray(cellValue)) {
          return (0, import_download.escapeForCSV)(cellValue);
        }
        try {
          const multiMeterBarValues = Object.fromEntries(
            cellValue.map((item) => [item.name, item.value])
          );
          return (0, import_download.escapeForCSV)(JSON.stringify(multiMeterBarValues));
        } catch {
          (0, import_logging.devLog)(
            `Failed to extract meterbar values for column "${column.id}". Expected array items to have 'name' and 'value' properties.`
          );
        }
      }
      if (typeof cellValue === "string" || Array.isArray(cellValue)) {
        return (0, import_download.escapeForCSV)(cellValue);
      }
      if (typeof cellValue === "object" && cellValue !== null) {
        if (Object.hasOwn(cellValue, "toString") && typeof cellValue.toString === "function") {
          return (0, import_download.escapeForCSV)(cellValue.toString());
        }
        try {
          return (0, import_download.escapeForCSV)(JSON.stringify(cellValue));
        } catch {
          (0, import_logging.devLog)(
            `Accessed cell value for column "${column.id}" received an object, but was not stringifyable. Consider providing a toString() function on the object.`
          );
        }
      }
      return `${cellValue}`;
    };
  },
  createTable(table) {
    table.downloadAllData = (excludeColumns, onDownloadEnd) => downloadData(
      table,
      table.getCoreRowModel(),
      "all",
      excludeColumns,
      onDownloadEnd
    );
    table.downloadPageData = (excludeColumns, onDownloadEnd) => downloadData(
      table,
      table.getPaginationRowModel(),
      "page",
      excludeColumns,
      onDownloadEnd
    );
    table.downloadSelectedData = (excludeColumns, onDownloadEnd) => downloadData(
      table,
      table.getSelectedRowModel(),
      "selected",
      excludeColumns,
      onDownloadEnd
    );
  }
};
function useDownloadData(props, options) {
  options.onDownloadData = props.onDownloadData;
}
