// @dynatrace/strato-components/esm/content/terminology-overlay/TerminologyOverlay.js
import { jsx, jsxs } from "react/jsx-runtime";
import { forwardRef } from "react";
import { Description } from "./Description.js";
import { Footer } from "./Footer.js";
import { COMPONENT_NAME } from "./terminology-overlay-constants.js";
import { TerminologyOverlayContext } from "./TerminologyOverlayContext.js";
import { Trigger } from "./Trigger.js";
import { useAriaLabelingProps } from "../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useId } from "../../core/hooks/useId.js";
import { useMergeRefs } from "../../core/hooks/useMergeRefs.js";
import { useSafeForwardProps } from "../../core/hooks/useSafeForwardProps.js";
import { useSlots } from "../../core/hooks/useSlots.js";
import { devLog } from "../../core/utils/logging.js";
import { Flex } from "../../layouts/flex/Flex.js";
import { useBreakpoint } from "../../layouts/hooks/useBreakpoint.js";
import { useOverlayWithTrigger } from "../../overlays/hooks/useOverlayWithTrigger.js";
import { Overlay } from "../../overlays/overlay/Overlay.js";
const TerminologyOverlayRoot = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const {
    children,
    defaultOpen,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
  const ariaLabelingProps = useAriaLabelingProps(restProps);
  const forwardProps = useSafeForwardProps(restProps, ariaLabelingProps);
  const smallerScreens = useBreakpoint("(max-width: 480px)");
  const overlayContentId = useId();
  const {
    targetRef,
    triggerRef,
    overlayRef,
    overlayState,
    overlayProps,
    overlayTriggerProps
  } = useOverlayWithTrigger({
    placement: "bottom-start",
    placementFallbacks: ["bottom-end", "top-start", "top-end"],
    offset: 4,
    crossOffset: -16,
    widthStrategy: smallerScreens ? { min: 200 } : { min: 350 },
    defaultOpen,
    shouldCloseOnBlur: true
  });
  const rootRef = useMergeRefs([forwardedRef, triggerRef, targetRef]);
  const [slots] = useSlots(children, {
    trigger: Trigger,
    description: Description,
    footer: { match: Footer, multiple: true }
  });
  if (slots.trigger === void 0) {
    devLog("Trigger element has not been defined!");
  }
  const { children: triggerChildren, ...triggerSlotProps } = slots.trigger?.props ?? {};
  const triggerProps = slots.trigger?.props ?? {};
  return /* @__PURE__ */ jsxs(TerminologyOverlayContext.Provider, { value: overlayTriggerProps, children: [
    /* @__PURE__ */ jsx(
      Trigger,
      {
        ref: rootRef,
        className: triggerProps.className,
        style: triggerProps.style,
        "aria-describedby": overlayState.isOpen ? overlayContentId : void 0,
        "data-testid": triggerProps["data-testid"],
        "data-dtrum-mask": triggerProps["data-dtrum-mask"] ?? dataDtrumMask,
        "data-dtrum-allow": triggerProps["data-dtrum-allow"] ?? dataDtrumAllow,
        ...triggerSlotProps,
        children: triggerChildren
      }
    ),
    /* @__PURE__ */ jsx(
      Overlay,
      {
        ref: overlayRef,
        isOpen: overlayState.isOpen,
        overlayProps,
        className: consumerClassName,
        style: consumerStyle,
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        ...ariaLabelingProps,
        ...behavioralTrackingProps,
        ...forwardProps,
        "data-dt-component": COMPONENT_NAME,
        children: /* @__PURE__ */ jsxs(
          Flex,
          {
            id: overlayContentId,
            paddingY: 12,
            paddingX: 16,
            flexDirection: "column",
            children: [
              slots.description,
              slots.footer
            ]
          }
        )
      }
    )
  ] });
});
TerminologyOverlayRoot.displayName = COMPONENT_NAME;
const TerminologyOverlay = Object.assign(TerminologyOverlayRoot, {
  Description,
  Trigger,
  Footer
});
export {
  TerminologyOverlay
};
