// @dynatrace/strato-components/esm/content/terminology-overlay/Trigger.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import { forwardRef, useContext } from "react";
import { COMPONENT_NAME } from "./terminology-overlay-constants.js";
import { trigger } from "./TerminologyOverlay.sty.js";
import { TerminologyOverlayContext } from "./TerminologyOverlayContext.js";
import { useAriaLabelingProps } from "../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useFocusRing } from "../../core/styles/useFocusRing.js";
import { mergeProps as _mergeProps } from "../../core/utils/merge-props.js";
const Trigger = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const {
    children,
    id: propId,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
  const {
    "data-dt-name": dataDtName = "Trigger",
    ...remainingBehavioralTrackingProps
  } = behavioralTrackingProps;
  const ariaLabelingProps = useAriaLabelingProps(restProps);
  const { focusProps, focusClassName, isFocusVisible } = useFocusRing({
    isMinimal: true
  });
  const overlayTriggerProps = useContext(TerminologyOverlayContext);
  return /* @__PURE__ */ jsx(
    "button",
    {
      className: clsx(
        focusClassName,
        trigger,
        {
          "strato-focus-visible": isFocusVisible
        },
        consumerClassName
      ),
      style: {
        textDecoration: isFocusVisible ? "none" : void 0,
        ...consumerStyle
      },
      ref: forwardedRef,
      ..._mergeProps(
        overlayTriggerProps,
        focusProps,
        ariaLabelingProps,
        remainingBehavioralTrackingProps
      ),
      "aria-haspopup": "dialog",
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "data-dt-name": dataDtName,
      "data-dt-component": COMPONENT_NAME,
      children
    }
  );
});
Trigger.displayName = "TerminologyOverlay.Trigger";
export {
  Trigger
};
