// @dynatrace/strato-components/charts/core/components/legend/range/components/RangeLegendSegments/RangeLegendSegments.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var RangeLegendSegments_exports = {};
__export(RangeLegendSegments_exports, {
  RangeLegendSegments: () => RangeLegendSegments
});
module.exports = __toCommonJS(RangeLegendSegments_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_constants = require("../../../../../constants.js");
var import_RangeLegendOrientationContext = require("../../context/RangeLegendOrientationContext.js");
var import_RangeLegendScaleContext = require("../../context/RangeLegendScaleContext.js");
var import_use_legend_formatted_labels = require("../../hooks/use-legend-formatted-labels.js");
var import_RangeLegendSegmentsItem = require("../RangeLegendSegmentsItem.js");
var import_RangeLegendSegments_css = require("./RangeLegendSegments.sty.js");
const RangeLegendSegments = (props) => {
  const { items } = props;
  const svgRef = (0, import_react.useRef)(null);
  const { scale } = (0, import_RangeLegendScaleContext.useScale)();
  const { orientation } = (0, import_RangeLegendOrientationContext.useOrientation)();
  const reversedScale = orientation === import_constants.ORIENTATION.Vertical;
  const internalRangeItems = (0, import_use_legend_formatted_labels.useLegendFormattedLabels)(items);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "svg",
    {
      ref: svgRef,
      className: (0, import_RangeLegendSegments_css.legendSvg)({
        isHorizontal: orientation === import_constants.ORIENTATION.Horizontal
      }),
      role: "presentation",
      "data-testid": "range-legend-segments",
      children: internalRangeItems.map((item) => {
        const startPosition = scale(reversedScale ? item.to : item.from);
        const endPosition = scale(reversedScale ? item.from : item.to);
        const segmentSize = endPosition - startPosition;
        return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_RangeLegendSegmentsItem.RangeLegendSegmentsItem,
          {
            item,
            segmentSize,
            startPosition
          },
          `segment-'${item.from}-${item.to}`
        );
      })
    }
  );
};
