// @dynatrace/strato-components/esm/charts/xy-chart-adoption/timeseries/utils/render-series.js
import { jsx } from "react/jsx-runtime";
import { getUnitAsString } from "../../../core/utils/get-unit-as-string.js";
import { XYChart } from "../../../xy-chart/XYChartRoot.js";
const xAxisId = "x-axis";
const defineAxisId = (unit) => `y-axis-${getUnitAsString(unit)}`;
const baseSeriesProps = ({ data, unit, seriesId }) => ({
  data,
  xAxisId,
  yAxisId: defineAxisId(unit),
  x0Accessor: "start",
  x1Accessor: "end",
  seriesIdAccessor: "seriesId",
  nameAccessor: "seriesName",
  key: seriesId
});
const renderLineSeries = (series) => /* @__PURE__ */ jsx(XYChart.LineSeries, { ...baseSeriesProps(series), y0Accessor: "value" });
const renderAreaSeries = (series) => /* @__PURE__ */ jsx(XYChart.AreaSeries, { ...baseSeriesProps(series), y1Accessor: "value" });
const renderBarSeries = (series) => /* @__PURE__ */ jsx(
  XYChart.BarSeries,
  {
    ...baseSeriesProps(series),
    variant: "stack",
    y1Accessor: "value"
  }
);
const renderBandSeries = (series) => /* @__PURE__ */ jsx(
  XYChart.AreaSeries,
  {
    ...baseSeriesProps(series),
    y0Accessor: "y0",
    y1Accessor: "y1"
  }
);
export {
  defineAxisId,
  renderAreaSeries,
  renderBandSeries,
  renderBarSeries,
  renderLineSeries,
  xAxisId
};
