// @dynatrace/strato-components/esm/charts/timeseries/components/displays/line/LineChartDisplay.js
import { jsx } from "react/jsx-runtime";
import { useMemo } from "react";
import {
  definedAccessor,
  xAxisAccessor as x,
  yAxisAccessor as y
} from "./accessors.js";
import { LineCanvas } from "../../../../core/components/canvas/components/LineCanvas.js";
import { shouldDisplayPoints } from "../../../../core/components/canvas/utils/display-points.js";
import { useGridPlotSize } from "../../../../core/components/chart-grid/hooks/useGridPlotSize.js";
import { isIsolatedPoint } from "../../../../core/utils/is-isolated-point.js";
import { safeGet } from "../../../../core/utils/safe-get.js";
import { LINE_CHART_DISPLAY_ARIA_LABEL } from "../../../constants.js";
import { useSeriesCurve } from "../../../hooks/useSeriesCurve.js";
import { useTimeseriesFilteredData } from "../../../hooks/useTimeseriesFilteredData.js";
import { useYScales } from "../../../hooks/useYScales.js";
import { getClampedTimeScale } from "../shared/utils/clamp-value.js";
const LineChartDisplay = ({ series, xScale }) => {
  const yScales = useYScales();
  const { seriesMap, datapointsMap } = useTimeseriesFilteredData();
  const curve = useSeriesCurve();
  const { width, height } = useGridPlotSize();
  const lineSeries = useMemo(() => [...series].reverse(), [series]);
  const isolatedAccessor = (displayPoints) => (datapoints, index) => displayPoints || isIsolatedPoint(
    datapoints,
    index,
    (point) => definedAccessor(point, datapointsMap)
  );
  const clampedScale = getClampedTimeScale(xScale);
  return /* @__PURE__ */ jsx("figure", { "aria-label": LINE_CHART_DISPLAY_ARIA_LABEL, children: lineSeries.map((series2) => {
    const seriesMetadata = safeGet(seriesMap, series2);
    const displayPoints = shouldDisplayPoints(
      seriesMetadata.pointsDisplay,
      width,
      series2.datapoints.length
    );
    return /* @__PURE__ */ jsx(
      LineCanvas,
      {
        data: series2.datapoints,
        canvasWidth: width,
        canvasHeight: height,
        x: x(xScale, datapointsMap, clampedScale),
        y: y(yScales, datapointsMap, seriesMetadata.unit),
        defined: (point) => definedAccessor(point, datapointsMap),
        displayScatterPlot: isolatedAccessor(displayPoints),
        color: () => seriesMetadata.color.Default,
        opacity: 1,
        curve,
        enableDotOverflow: true
      },
      seriesMetadata.id
    );
  }) });
};
LineChartDisplay["displayName"] = "LineChartDisplay";
export {
  LineChartDisplay
};
