// @dynatrace/strato-components/charts/core/components/crosshair/components/Magnifier.sty.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Magnifier_css_exports = {};
__export(Magnifier_css_exports, {
  bottom: () => bottom,
  left: () => left,
  magnifier: () => magnifier,
  magnifierFontStyles: () => magnifierFontStyles,
  right: () => right,
  rotatedBottomStyle: () => rotatedBottomStyle,
  rotatedTopStyle: () => rotatedTopStyle,
  top: () => top
});
module.exports = __toCommonJS(Magnifier_css_exports);
var import_Magnifier_css_ts_vanilla = require("./Magnifier.css");
var bottom = "_nm3s3e3-3-6-0";
var left = "_nm3s3e1-3-6-0";
var magnifier = "_nm3s3e0-3-6-0";
var magnifierFontStyles = { fontFamily: "var(--dt-typography-text-small-default-family, DynatraceFlow,Roboto,Helvetica,sans-serif)", fontSize: "var(--dt-typography-text-small-default-size, 12px)", fontWeight: "var(--dt-typography-text-small-default-weight, 400)", lineHeight: "var(--dt-typography-text-small-default-line-height, 1.333333333)" };
var right = "_nm3s3e2-3-6-0";
var rotatedBottomStyle = "_nm3s3e6-3-6-0";
var rotatedTopStyle = "_nm3s3e5-3-6-0";
var top = "_nm3s3e4-3-6-0";
