// @dynatrace/strato-components/charts/honeycomb/components/canvas/renderers/CanvasCircleRenderer.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CanvasCircleRenderer_exports = {};
__export(CanvasCircleRenderer_exports, {
  CanvasCircleRenderer: () => CanvasCircleRenderer
});
module.exports = __toCommonJS(CanvasCircleRenderer_exports);
var import_react = require("react");
var import_useCanvas = require("../../../../core/components/canvas/hooks/useCanvas.js");
var import_draw_label_and_value = require("../utils/draw-label-and-value.js");
var import_fill_shape = require("../utils/fill-shape.js");
var import_remove_transparent_border = require("../utils/remove-transparent-border.js");
const createCirclePath = (ctx, radius, x, y) => {
  ctx.beginPath();
  const safeRadius = Math.max(radius, 0);
  ctx.arc(x, y, safeRadius, 0, 2 * Math.PI);
  ctx.closePath();
};
const getCircleTextArea = (x, y, squareSideSize) => ({
  x,
  y,
  width: squareSideSize,
  height: squareSideSize - 8,
  leftBoundary: x - squareSideSize / 2,
  rightBoundary: x + squareSideSize / 2
});
const getCircumscribedSquare = (radius) => radius * Math.sqrt(2);
const CanvasCircleRenderer = (props) => {
  const {
    x,
    y,
    radius,
    fillColor,
    strokeColor = fillColor,
    value,
    name = "",
    showLabels,
    truncationMode
  } = props;
  const squareSideSize = getCircumscribedSquare(radius);
  const textArea = getCircleTextArea(x, y, squareSideSize);
  const draw = (0, import_react.useCallback)(
    (ctx) => {
      createCirclePath(ctx, radius, x, y);
      (0, import_remove_transparent_border.removeTransparentBorder)(ctx, strokeColor);
      (0, import_fill_shape.fillShape)(ctx, fillColor);
      if (showLabels) {
        (0, import_draw_label_and_value.drawLabelAndValue)(
          ctx,
          value,
          textArea,
          fillColor,
          name,
          truncationMode
        );
      }
      return false;
    },
    [
      x,
      y,
      radius,
      fillColor,
      strokeColor,
      showLabels,
      truncationMode,
      name,
      value
    ]
  );
  (0, import_useCanvas.useCanvas)("honeycomb-circle", draw);
  return null;
};
CanvasCircleRenderer["displayName"] = "CanvasCircleRenderer";
