// @dynatrace/strato-components/charts/honeycomb/components/canvas/renderers/CanvasSquareRenderer.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CanvasSquareRenderer_exports = {};
__export(CanvasSquareRenderer_exports, {
  CanvasSquareRenderer: () => CanvasSquareRenderer
});
module.exports = __toCommonJS(CanvasSquareRenderer_exports);
var import_react = require("react");
var import_useCanvas = require("../../../../core/components/canvas/hooks/useCanvas.js");
var import_draw_label_and_value = require("../utils/draw-label-and-value.js");
var import_fill_shape = require("../utils/fill-shape.js");
var import_remove_transparent_border = require("../utils/remove-transparent-border.js");
const createRoundedSquarePath = (ctx, x, y, width, radius) => {
  ctx.moveTo(x, y);
  ctx.beginPath();
  ctx.roundRect(x, y, width, width, radius);
};
const getSquareTextArea = (x, y, width) => ({
  x: x + width / 2,
  y: y + width / 2,
  width,
  height: width,
  leftBoundary: x,
  rightBoundary: x + width
});
const CanvasSquareRenderer = (props) => {
  const {
    x,
    y,
    width,
    fillColor,
    strokeColor = fillColor,
    value,
    name = "",
    radius = 0,
    showLabels,
    truncationMode
  } = props;
  const textArea = getSquareTextArea(x, y, width);
  const draw = (0, import_react.useCallback)(
    (ctx) => {
      createRoundedSquarePath(ctx, x, y, width, radius);
      (0, import_remove_transparent_border.removeTransparentBorder)(ctx, strokeColor);
      (0, import_fill_shape.fillShape)(ctx, fillColor);
      if (showLabels) {
        (0, import_draw_label_and_value.drawLabelAndValue)(
          ctx,
          value,
          textArea,
          fillColor,
          name,
          truncationMode
        );
      }
      return false;
    },
    [
      x,
      y,
      width,
      fillColor,
      strokeColor,
      showLabels,
      truncationMode,
      name,
      value,
      radius
    ]
  );
  (0, import_useCanvas.useCanvas)("honeycomb-square", draw);
  return null;
};
CanvasSquareRenderer["displayName"] = "CanvasSquareRenderer";
