// @dynatrace/strato-components/overlays/drawer/Drawer.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Drawer_exports = {};
__export(Drawer_exports, {
  Drawer: () => Drawer
});
module.exports = __toCommonJS(Drawer_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_framer_motion = require("framer-motion");
var import_react = require("react");
var import_animations_js = __toESM(require("@dynatrace/strato-design-tokens/animations-js"));
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
var import_Drawer_css = require("./Drawer.sty.js");
var import_FocusScope = require("../../core/components/focus-scope/FocusScope.js");
var import_OverlayContainer = require("../../core/components/OverlayContainer.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_useOverlayStacking = require("../../core/hooks/useOverlayStacking.js");
var import_OverlayStackingProvider = require("../../core/providers/OverlayStackingProvider.js");
var import_framer_motion_easings = require("../../core/utils/framer-motion-easings.js");
var import_useOverlay = require("../hooks/useOverlay.js");
const COMPONENT_NAME = "Drawer";
const slideAnimation = ($placement) => ({
  visible: {
    x: "0",
    transition: {
      duration: import_animations_js.default.SlideIn.Duration.Slow,
      ease: import_framer_motion_easings.slideInEasing
    },
    // Workaround to fix remaining style (transform: translateX(0)) on Drawer
    // as this breaks any child element with position: fixed
    // https://github.com/framer/motion/issues/823
    transitionEnd: {
      x: 0
    }
  },
  hidden: {
    x: $placement === "left" ? "-100%" : "100%",
    transition: {
      duration: import_animations_js.default.SlideOut.Duration.Slow,
      ease: import_framer_motion_easings.slideOutEasing
    }
  }
});
const fade = {
  fadeIn: {
    // For the opacity we need to use the default value here. Framer motion would complain
    // about too complex values, when transitioning from two custom property values within
    // an rgba value apparently.
    opacity: 1,
    transition: {
      duration: import_animations_js.default.SlideIn.Duration.Slow,
      ease: import_framer_motion_easings.fadeInEasing
    }
  },
  fadeOut: {
    opacity: 0,
    transition: {
      duration: import_animations_js.default.SlideOut.Duration.Slow,
      ease: import_framer_motion_easings.fadeOutEasing
    }
  }
};
const Drawer = /* @__PURE__ */ (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const {
      placement = "right",
      width = "medium",
      isDismissed,
      onDismiss,
      children,
      className: consumerClassName,
      style: consumerStyle,
      modal = true,
      "data-testid": dataTestId,
      disableAnimations = false,
      ...remainingProps
    } = props;
    const ref = (0, import_react.useRef)(null);
    const { overlayProps } = (0, import_useOverlay.useOverlay)(
      {
        onClose: (reason) => {
          onDismiss(true, reason);
        },
        isOpen: !isDismissed,
        isDismissible: false,
        isKeyboardDismissDisabled: false
      },
      ref
    );
    const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
    const mergedRefs = (0, import_useMergeRefs.useMergeRefs)([ref, forwardedRef]);
    const overlayContainerProps = (0, import_useOverlayStacking.useOverlayStacking)(0, "overlay");
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_framer_motion.AnimatePresence, { children: !isDismissed && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_OverlayContainer.OverlayContainer,
      {
        "data-testid": "drawer-container",
        ...overlayContainerProps,
        style: {
          ...overlayContainerProps.style,
          pointerEvents: modal ? void 0 : "none"
        },
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          DrawerBackdrop,
          {
            modal,
            overlayProps,
            onDismiss,
            disableAnimations,
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_framer_motion.motion.div,
              {
                "data-testid": dataTestId,
                ref: mergedRefs,
                variants: disableAnimations ? {} : slideAnimation(placement),
                initial: "hidden",
                animate: "visible",
                exit: "hidden",
                onAnimationComplete: () => window.dispatchEvent(new Event("animationDone")),
                tabIndex: -1,
                className: (0, import_clsx.default)(
                  consumerClassName,
                  import_Drawer_css.DrawerCSS,
                  import_Drawer_css.DrawerPlacementCSS[placement],
                  {
                    [import_Drawer_css.DrawerWidthStyles.narrow]: width === "narrow",
                    [import_Drawer_css.DrawerWidthStyles.medium]: width === "medium",
                    [import_Drawer_css.DrawerWidthStyles.wide]: width === "wide",
                    [import_Drawer_css.DrawerWidthStyles.full]: width === "full"
                  }
                ),
                style: {
                  width: (
                    // Avoid binding the predefined named widths to the style, but add any other values
                    width !== "narrow" && width !== "medium" && width !== "wide" && width !== "full" ? width : void 0
                  ),
                  ...consumerStyle
                },
                ...restProps,
                ...behavioralTrackingProps,
                "data-dt-component": COMPONENT_NAME,
                children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_FocusScope.FocusScope, { autoFocus: true, contain: modal, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  import_framer_motion.motion.div,
                  {
                    variants: disableAnimations ? {} : fade,
                    initial: "fadeOut",
                    animate: "fadeIn",
                    exit: "fadeOut",
                    className: import_Drawer_css.DrawerContentCSS,
                    children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayStackingProvider.OverlayStackingProvider, { children })
                  }
                ) })
              }
            )
          }
        )
      }
    ) });
  }
);
const DrawerBackdrop = (props) => {
  const { children, modal, disableAnimations, onDismiss, overlayProps } = props;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_framer_motion.motion.div,
    {
      variants: disableAnimations ? {} : fade,
      initial: "fadeOut",
      animate: "fadeIn",
      exit: "fadeOut",
      "data-testid": "drawer-backdrop",
      className: import_Drawer_css.DrawerBackdropCSS,
      style: {
        pointerEvents: modal ? void 0 : "none",
        backgroundColor: modal ? import_colors.default.Background.Surface.Backdrop : void 0
      },
      onPointerDown: (event) => {
        if (modal === true && event.target === event.currentTarget) {
          onDismiss(true, "backdrop");
        }
      },
      ...overlayProps,
      children
    }
  );
};
