// @dynatrace/strato-components/charts/timeseries/components/legend/TimeseriesChartLegend.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TimeseriesChartLegend_exports = {};
__export(TimeseriesChartLegend_exports, {
  TimeseriesChartLegend: () => TimeseriesChartLegend,
  renderIcon: () => renderIcon
});
module.exports = __toCommonJS(TimeseriesChartLegend_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
var import_strato_icons = require("@dynatrace/strato-icons");
var import_useChartLayoutLoading = require("../../../core/components/chart-layout/hooks/useChartLayoutLoading.js");
var import_useChartLayoutOrientation = require("../../../core/components/chart-layout/hooks/useChartLayoutOrientation.js");
var import_orientation = require("../../../core/components/chart-layout/utils/orientation.js");
var import_explore_reducers = require("../../../core/components/interactions-state-helpers/reducers/explore-reducers.js");
var import_ChartLegend = require("../../../core/components/legend/categorical/ChartLegend.js");
var import_useLegendSeriesActions = require("../../../core/components/legend/categorical/hooks/useLegendSeriesActions.js");
var import_safe_get = require("../../../core/utils/safe-get.js");
var import_single_line_name = require("../../../core/utils/single-line-name.js");
var import_selectors = require("../../hooks/store/selectors.js");
var import_useTimeseriesData = require("../../hooks/useTimeseriesData.js");
var import_useTruncationMode = require("../../hooks/useTruncationMode.js");
var import_timeseries2 = require("../../utils/timeseries.js");
const renderIcon = (variant) => {
  switch (variant) {
    case "bar":
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.BarChartIcon, { size: "small" });
    case "area":
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.StackedAreaChartIcon, { size: "small" });
    case "band":
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.BandChartIcon, { size: "small" });
    case "line":
    default:
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.LineChartIcon, { size: "small" });
  }
};
const createItemData = (series, seriesMap) => {
  const isMultiGeometries = (0, import_timeseries2.hasMultiGeometries)(series);
  const result = [];
  for (const variant of series.shapePriority) {
    const variantSeries = series[variant];
    for (let index = 0; index < variantSeries.length; index++) {
      const singleSeries = variantSeries[index];
      const seriesMetadata = (0, import_safe_get.safeGet)(seriesMap, singleSeries);
      result.push({
        id: seriesMetadata.id ?? `item-${index}`,
        color: seriesMetadata.color ?? import_colors.default.Text.Neutral.Default,
        icon: isMultiGeometries ? renderIcon(variant) : void 0,
        name: (0, import_single_line_name.singleLineName)(singleSeries.name) ?? seriesMetadata.id
      });
    }
  }
  return result;
};
const TimeseriesChartLegend = (0, import_react.memo)(() => {
  const { dataPerVariant, data, seriesMap } = (0, import_useTimeseriesData.useTimeseriesData)();
  const truncationMode = (0, import_useTruncationMode.useTruncationMode)();
  const setState = (0, import_selectors.useSetState)();
  const { orientation: layoutOrientation } = (0, import_useChartLayoutOrientation.useChartLayoutOrientation)();
  const isLoading = (0, import_useChartLayoutLoading.useChartLayoutLoading)();
  const legendOrientation = (0, import_orientation.getLegendOrientation)(layoutOrientation);
  const legendActions = (0, import_useLegendSeriesActions.useLegendSeriesActions)(
    data,
    (s) => (0, import_safe_get.safeGet)(seriesMap, s).id
  );
  const legendItems = (0, import_react.useMemo)(
    () => createItemData(dataPerVariant, seriesMap),
    [dataPerVariant, seriesMap]
  );
  const updateHighlightedSeries = (item) => {
    setState((prev) => ({
      ...prev,
      highlightedSeriesId: item?.id
    }));
  };
  const updateVisibleSeries = (visibleSeries) => {
    const visibleSeriesSet = new Set(visibleSeries.map((s) => s.id));
    setState(
      (prev) => (0, import_explore_reducers.reduceExploreCancelTooltip)({
        ...prev,
        filteredSeriesIds: visibleSeriesSet
      })
    );
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ChartLegend.ChartLegend,
    {
      items: legendItems,
      orientation: legendOrientation,
      interactionsDisabled: isLoading,
      onHover: updateHighlightedSeries,
      onClick: updateVisibleSeries,
      onSeriesAction: legendActions,
      truncationMode
    }
  );
});
TimeseriesChartLegend["displayName"] = "TimeseriesChartLegend";
