// @dynatrace/strato-components/charts/histogram/components/axes/HistogramValueLeftAxisArea.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var HistogramValueLeftAxisArea_exports = {};
__export(HistogramValueLeftAxisArea_exports, {
  HistogramValueLeftAxisArea: () => HistogramValueLeftAxisArea
});
module.exports = __toCommonJS(HistogramValueLeftAxisArea_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_ValueLeftAxis = require("../../../core/components/axes/components/ValueLeftAxis.js");
var import_ChartGridRoot = require("../../../core/components/chart-grid/ChartGridRoot.js");
var import_histogram_grid_area_rules = require("../../constants/histogram-grid-area-rules.js");
var import_useHistogramTickAxis = require("../../hooks/useHistogramTickAxis.js");
var import_useHistogramYScales = require("../../hooks/useHistogramYScales.js");
var import_useYAxisConfig = require("../../hooks/useYAxisConfig.js");
const HistogramValueLeftAxisArea = () => {
  const { left: axisLeft } = (0, import_useYAxisConfig.useYAxisConfig)();
  const ticks = (0, import_useHistogramTickAxis.useHistogramTickAxis)();
  const [yScale] = (0, import_useHistogramYScales.useHistogramYScales)();
  const leftTicks = ticks?.left || { ticks: [], formattedTicks: [] };
  if ((0, import_lodash_es.isEmpty)(leftTicks.ticks) || !yScale) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ChartGridRoot.ChartGrid.GridArea, { gridArea: import_histogram_grid_area_rules.AREAS.leftAxis, children: ({ availableWidth, availableHeight }) => {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_ValueLeftAxis.ValueLeftAxis,
      {
        ticks: leftTicks.ticks,
        formattedTicks: leftTicks.formattedTicks,
        containerHeight: availableHeight,
        containerWidth: availableWidth,
        scale: yScale,
        label: axisLeft?.label
      }
    );
  } });
};
HistogramValueLeftAxisArea["displayName"] = "HistogramValueLeftAxisArea";
