// @dynatrace/strato-components/charts/pie/components/PieChartToolbar/PieChartToolbar.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var PieChartToolbar_exports = {};
__export(PieChartToolbar_exports, {
  PieChartToolbar: () => PieChartToolbar
});
module.exports = __toCommonJS(PieChartToolbar_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_DownloadDataButton = require("./components/DownloadDataButton.js");
var import_useBehavioralTrackingProps = require("../../../../core/hooks/useBehavioralTrackingProps.js");
var import_useTooltipReducer = require("../../../core/components/overlay-tooltip/useTooltipReducer.js");
var import_ChartToolbarRenderer = require("../../../core/components/toolbar/ChartToolbarRenderer.js");
var import_PieChartRawDataContext = require("../../context/PieChartRawDataContext.js");
var import_ToolbarContext = require("../../context/ToolbarContext.js");
var import_selectors = require("../../hooks/store/selectors.js");
var import_default = require("../../types/default.js");
var import_build_and_download_csv = require("../../utils/build-and-download-csv.js");
const defaultToolbarPlacement = "top-right";
const PieChartToolbar = () => {
  const dispatch = (0, import_useTooltipReducer.useTooltipReducer)();
  const toolbarConfig = (0, import_ToolbarContext.useToolbarConfig)();
  const toolbarPlacement = toolbarConfig?.placement ?? defaultToolbarPlacement;
  const { mouseInBounds } = (0, import_selectors.useOverlayState)();
  const isDownloadEnabled = toolbarConfig?.downloadData?.enabled ?? false;
  const rawData = (0, import_PieChartRawDataContext.usePieChartRawData)();
  const handleDownloadDataToolbar = () => (0, import_build_and_download_csv.buildAndDownloadCsv)(rawData);
  const showDefaultToolbarDownloadButton = isDownloadEnabled && !toolbarConfig?.intents;
  const showMenuToolbarButton = isDownloadEnabled && !!toolbarConfig?.intents?.length;
  const initialA11yState = showDefaultToolbarDownloadButton ? [{ download: 0 }] : [{ intent: 0 }];
  const behaviorTrackingProps = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(
    toolbarConfig ?? {}
  )[0];
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ChartToolbarRenderer.ChartToolbarRenderer,
    {
      visible: mouseInBounds,
      placement: toolbarPlacement,
      collapsed: toolbarConfig?.collapsed,
      draggable: true,
      zIndex: import_default.zIndicesPieAreas.toolbar,
      initialA11yState,
      onMouseEnter: dispatch({ type: "RESET_TOOLTIP" }),
      menuDisplayMode: "never",
      isDownloadEnabled: showMenuToolbarButton,
      intents: toolbarConfig?.intents,
      handleDownloadDataToolbar,
      removeDragHandle: true,
      ...behaviorTrackingProps,
      children: showDefaultToolbarDownloadButton && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_DownloadDataButton.DownloadDataButton, {})
    }
  );
};
