// @dynatrace/strato-components/esm/filters/filter-field/hooks/useSuggestions.js
import { EditorView } from "@codemirror/view";
import {
  isValidElement,
  useCallback,
  useMemo,
  useRef,
  useState
} from "react";
import { useErrorBoundary } from "react-error-boundary";
import { useIntl } from "react-intl";
import { useErrorLogsStore } from "./useErrorLogsStore.js";
import { useReferencedState } from "./useReferencedState.js";
import { usePrevious } from "../../../core/hooks/usePrevious.js";
import { useFormFieldValidation } from "../../../forms/forms-core/hooks/useFormFieldValidation.js";
import {
  ANNOTATION_SUGGESTION_APPLY,
  ariaAttributesCompartment,
  stringAnnotation
} from "../editor-state-configs/editor-config.js";
import { filterFieldMessages } from "../messages.js";
import {
  getCurrentFocusableListItem
} from "../utils/get-next-suggestion.js";
import { getSuggestionTransaction } from "../utils/get-suggestion-transaction.js";
function stringifySuggestions(value) {
  return JSON.stringify(value, replacer);
}
function replacer(_key, value) {
  if (isValidElement(value)) {
    const type = value.type;
    return `$$ReactElement:${typeof type === "string" ? type : type.displayName ?? type.name ?? "Component"}`;
  }
  return value;
}
function useSuggestions(props) {
  const {
    ariaDisabled,
    autoSuggestions,
    isOpen,
    id,
    suggestionsOverlayRef,
    editorViewRef,
    onSuggest,
    viewDispatch,
    hiddenInputRef,
    ariaLabelingProps
  } = props;
  const { showBoundary } = useErrorBoundary();
  const setUserAction = useErrorLogsStore((state) => state.setAction);
  const { ariaProps } = useFormFieldValidation(
    hiddenInputRef,
    {},
    ariaLabelingProps
  );
  const [
    autoSuggestionValues,
    setAutoSuggestionValues,
    autoSuggestionValuesRef
  ] = useReferencedState({});
  const [
    recentSuggestionValues,
    setRecentSuggestionValues,
    recentSuggestionValuesRef
  ] = useReferencedState([]);
  const [
    pinnedSuggestionValues,
    setPinnedSuggestionValues,
    pinnedSuggestionValuesRef
  ] = useReferencedState([]);
  const [
    logicalOperatorSuggestionValues,
    setLogicalOperatorSuggestionValues,
    logicalOperatorSuggestionValuesRef
  ] = useReferencedState([]);
  const [
    pastedContentSuggestionValues,
    setPastedContentSuggestionValues,
    pastedContentSuggestionValuesRef
  ] = useReferencedState([]);
  const [
    searchSuggestionValues,
    setSearchSuggestionValues,
    searchSuggestionValuesRef
  ] = useReferencedState([]);
  const [
    fallbackKeySuggestionValues,
    setFallbackKeySuggestionValues,
    fallbackKeySuggestionValuesRef
  ] = useReferencedState([]);
  const [
    statementSuggestionValues,
    setStatementSuggestionValues,
    statementSuggestionValuesRef
  ] = useReferencedState([]);
  const [prevAutoSuggestionValues, setPrevAutoSuggestionValues] = useState({});
  const [prevRecentSuggestionValues, setPrevRecentSuggestionValues] = useState([]);
  const [
    prevPastedContentSuggestionValues,
    setPrevPastedContentSuggestionValues
  ] = useState([]);
  const [prevSearchSuggestionValues, setPrevSearchSuggestionValues] = useState([]);
  const [prevFallbackKeySuggestionValues, setPrevFallbackKeySuggestionValues] = useState([]);
  const [prevStatementSuggestionValues, setPrevStatementSuggestionValues] = useState([]);
  const [focusedListItem, setFocusedListItem] = useState();
  const prevFocusedListItem = usePrevious(focusedListItem);
  const suggestionsContext = useRef(
    void 0
  );
  const _onSuggest = useCallback(
    (args) => {
      suggestionsContext.current = args.suggestionsContext;
      if (autoSuggestions && stringifySuggestions(args.autoSuggestionValues) !== stringifySuggestions(autoSuggestionValuesRef.current)) {
        setAutoSuggestionValues(args.autoSuggestionValues);
      }
      if (stringifySuggestions(args.recentSuggestionValues) !== stringifySuggestions(recentSuggestionValuesRef.current)) {
        setRecentSuggestionValues(args.recentSuggestionValues);
      }
      if (stringifySuggestions(args.pinnedSuggestionValues) !== stringifySuggestions(pinnedSuggestionValuesRef.current)) {
        setPinnedSuggestionValues(args.pinnedSuggestionValues);
      }
      if (stringifySuggestions(args.logicalOperatorSuggestionValues) !== stringifySuggestions(logicalOperatorSuggestionValuesRef.current)) {
        setLogicalOperatorSuggestionValues(
          args.logicalOperatorSuggestionValues
        );
      }
      if (stringifySuggestions(args.pastedContentSuggestionValues) !== stringifySuggestions(pastedContentSuggestionValuesRef.current)) {
        setPastedContentSuggestionValues(args.pastedContentSuggestionValues);
      }
      if (stringifySuggestions(args.searchSuggestionValues) !== stringifySuggestions(searchSuggestionValuesRef.current)) {
        setSearchSuggestionValues(args.searchSuggestionValues);
      }
      if (stringifySuggestions(args.fallbackKeySuggestionValues) !== stringifySuggestions(fallbackKeySuggestionValuesRef.current)) {
        setFallbackKeySuggestionValues(args.fallbackKeySuggestionValues);
      }
      if (stringifySuggestions(args.filterStatementSuggestionValues) !== stringifySuggestions(statementSuggestionValuesRef.current)) {
        setStatementSuggestionValues(args.filterStatementSuggestionValues);
      }
    },
    [
      autoSuggestions,
      autoSuggestionValuesRef,
      recentSuggestionValuesRef,
      pinnedSuggestionValuesRef,
      logicalOperatorSuggestionValuesRef,
      pastedContentSuggestionValuesRef,
      searchSuggestionValuesRef,
      fallbackKeySuggestionValuesRef,
      statementSuggestionValuesRef,
      setAutoSuggestionValues,
      setRecentSuggestionValues,
      setPinnedSuggestionValues,
      setLogicalOperatorSuggestionValues,
      setPastedContentSuggestionValues,
      setSearchSuggestionValues,
      setFallbackKeySuggestionValues,
      setStatementSuggestionValues
    ]
  );
  const [prevOpen, setPrevOpen] = useState(isOpen);
  const intl = useIntl();
  const contentEditableDivAriaLabel = intl.formatMessage(
    filterFieldMessages.ariaLabel
  );
  const suggestionsOverlayHintId = "suggestions-overlay-announcement";
  const ariaDescribedBys = ariaProps["aria-describedby"] ?? "";
  const suggestionsHintId = isOpen ? suggestionsOverlayHintId : "";
  const ariaLabelProp = ariaLabelingProps?.["aria-label"];
  const ariaDetailsProp = ariaLabelingProps?.["aria-details"] ?? "";
  const ariaLabelledbyProp = ariaLabelingProps?.["aria-labelledby"] ?? "";
  const ariaAttributes = useMemo(
    () => ({
      "aria-label": ariaLabelProp ?? contentEditableDivAriaLabel,
      "aria-autocomplete": "list",
      "aria-haspopup": "listbox",
      ...isOpen && {
        "aria-controls": `suggestions-${id}`
      },
      ...focusedListItem?.id && {
        "aria-activedescendant": focusedListItem?.id
      },
      ...ariaDisabled && {
        "aria-disabled": `${ariaDisabled}`
      },
      "aria-describedby": (ariaDescribedBys + " " + suggestionsHintId).trim(),
      ...ariaDetailsProp && {
        "aria-details": `${ariaDetailsProp}`
      },
      ...ariaLabelledbyProp && {
        "aria-labelledby": `${ariaLabelledbyProp}`
      }
    }),
    [
      ariaDescribedBys,
      ariaDisabled,
      ariaLabelProp,
      ariaDetailsProp,
      ariaLabelledbyProp,
      contentEditableDivAriaLabel,
      focusedListItem,
      isOpen,
      suggestionsHintId,
      id
    ]
  );
  const applySuggestion = useCallback(
    (element, value, trailing, insertionStrategy, cursorBeforeTrailing, range, shouldEscapeSuggestion) => {
      if (!editorViewRef.current || !suggestionsContext.current || !element) {
        return;
      }
      setUserAction({
        type: "applySuggestion",
        data: {
          value,
          trailing,
          insertionStrategy,
          cursorBeforeTrailing
        }
      });
      const suggestionTransaction = getSuggestionTransaction(
        value,
        range,
        shouldEscapeSuggestion,
        trailing,
        insertionStrategy,
        cursorBeforeTrailing,
        suggestionsContext.current,
        editorViewRef.current.state
      );
      viewDispatch({
        ...suggestionTransaction,
        annotations: stringAnnotation.of(ANNOTATION_SUGGESTION_APPLY)
      });
      setFocusedListItem(void 0);
    },
    // Excluding suggestionsContext. We are accessing the current context by reference, will always be up to date.
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [setFocusedListItem, viewDispatch]
  );
  if (prevOpen && !isOpen) {
    queueMicrotask(() => {
      try {
        onSuggest?.();
      } catch (e) {
        showBoundary(e);
      }
    });
  }
  if (prevOpen !== isOpen || autoSuggestionValuesRef.current !== prevAutoSuggestionValues || recentSuggestionValuesRef.current !== prevRecentSuggestionValues || pastedContentSuggestionValuesRef.current !== prevPastedContentSuggestionValues || searchSuggestionValuesRef.current !== prevSearchSuggestionValues || fallbackKeySuggestionValuesRef.current !== prevFallbackKeySuggestionValues || statementSuggestionValuesRef.current !== prevStatementSuggestionValues) {
    if (isOpen) {
      if (!getCurrentFocusableListItem(
        suggestionsOverlayRef.current,
        focusedListItem?.id
      )) {
        setFocusedListItem(void 0);
      }
    }
    setPrevAutoSuggestionValues(autoSuggestionValuesRef.current ?? {});
    setPrevRecentSuggestionValues(recentSuggestionValuesRef.current ?? []);
    setPrevPastedContentSuggestionValues(
      pastedContentSuggestionValuesRef.current ?? []
    );
    setPrevSearchSuggestionValues(searchSuggestionValuesRef.current ?? []);
    setPrevFallbackKeySuggestionValues(
      fallbackKeySuggestionValuesRef.current ?? []
    );
    setPrevStatementSuggestionValues(
      statementSuggestionValuesRef.current ?? []
    );
    setPrevOpen(isOpen);
  }
  if (prevFocusedListItem !== focusedListItem || prevOpen !== isOpen) {
    viewDispatch({
      effects: ariaAttributesCompartment.reconfigure(
        EditorView.contentAttributes.of(ariaAttributes)
      )
    });
  }
  const handleNextSuggestion = (nextSuggestion) => {
    if (nextSuggestion) {
      const id2 = nextSuggestion.dataset.stratoFilterFieldVirtualFocusId;
      const type = nextSuggestion.dataset.stratoFilterFieldVirtualFocusType;
      setFocusedListItem({ id: id2, type });
      nextSuggestion.scrollIntoView({ block: "center" });
    }
  };
  const autoSuggestionsConfig = useMemo(() => {
    const filterValue = suggestionsContext.current?.currentStatement?.value;
    const cursorPosition = editorViewRef.current?.state.selection.main.head;
    return {
      trailing: cursorPosition === filterValue?.range.to ? " " : void 0,
      groupedSuggestions: autoSuggestionValues
    };
  }, [autoSuggestionValues]);
  return {
    ariaAttributes,
    autoSuggestionsConfig,
    recentSuggestionValues,
    pinnedSuggestionValues,
    logicalOperatorSuggestionValues,
    pastedContentSuggestionValues,
    searchSuggestionValues,
    fallbackKeySuggestionValues,
    statementSuggestionValues,
    focusedListItem,
    suggestionsContext,
    applySuggestion,
    handleNextSuggestion,
    _onSuggest,
    suggestionsOverlayHintId
  };
}
export {
  stringifySuggestions,
  useSuggestions
};
