// @dynatrace/strato-components/esm/filters/filter-field/hooks/useShowMore.js
import { jsx, jsxs } from "react/jsx-runtime";
import { useLayoutEffect, useMemo, useState } from "react";
import { FormattedMessage } from "react-intl";
import {
  ChevronDownSmallIcon,
  ChevronUpSmallIcon
} from "@dynatrace/strato-icons";
import { useId } from "../../../core/hooks/useId.js";
import { getVirtualFocusStyles } from "../../../core/styles/getVirtualFocusStyles.js";
import { Flex } from "../../../layouts/flex/Flex.js";
import { filterFieldMessages } from "../messages.js";
import { useFilterFieldContext } from "./useFilterFieldContext.js";
import {
  getCurrentFocusableListItemByIndex,
  getCurrentFocusableListItemIndex
} from "../utils/get-next-suggestion.js";
function useShowMore(props) {
  const { items, primaryCount, sliceFn: _propSliceFn } = props;
  const sliceFn = _propSliceFn ?? defaultSlice;
  const [isExpanded, setIsExpanded] = useState(false);
  const [pendingNextSuggestionIndex, setPendingNextSuggestionIndex] = useState(null);
  const { focusedListItem, handleNextSuggestion, suggestionsOverlayRef } = useFilterFieldContext();
  const id = useId();
  useLayoutEffect(() => {
    if (pendingNextSuggestionIndex !== null) {
      const element = getCurrentFocusableListItemByIndex(
        suggestionsOverlayRef,
        pendingNextSuggestionIndex
      );
      handleNextSuggestion(element);
      setPendingNextSuggestionIndex(null);
    }
  }, [pendingNextSuggestionIndex, suggestionsOverlayRef, handleNextSuggestion]);
  const controlProps = useMemo(
    () => ({
      // Role needs to be option, because otherwise NVDA will crash if the activeDecendant of a listbox is anything other than a role option or group.
      role: "option",
      onClick: (event) => {
        event.preventDefault();
        event.stopPropagation();
        const isKeyboard = event.detail === 0 || event.nativeEvent instanceof KeyboardEvent;
        if (isExpanded || !isKeyboard) {
          setIsExpanded((previousState) => !previousState);
          return;
        }
        const index = getCurrentFocusableListItemIndex(
          suggestionsOverlayRef,
          focusedListItem?.id
        );
        setIsExpanded((previousState) => !previousState);
        setPendingNextSuggestionIndex(index ?? null);
      },
      id,
      "data-strato-filter-field-virtual-focus-id": id,
      "data-strato-filter-field-virtual-focus-type": "button",
      "data-dt-name": "ShowMoreLess",
      "data-dt-properties": `state:${isExpanded ? "less" : "more"}`,
      className: getVirtualFocusStyles(id === focusedListItem?.id).className,
      width: "full",
      textAlign: "start",
      children: /* @__PURE__ */ jsxs(Flex, { gap: 2, as: "span", paddingLeft: 4, children: [
        /* @__PURE__ */ jsx(
          FormattedMessage,
          {
            ...filterFieldMessages.suggestions.showMore,
            values: { isExpanded }
          }
        ),
        isExpanded ? /* @__PURE__ */ jsx(ChevronUpSmallIcon, {}) : /* @__PURE__ */ jsx(ChevronDownSmallIcon, {})
      ] })
    }),
    [focusedListItem?.id, id, isExpanded, suggestionsOverlayRef]
  );
  const { items: resultItems, isSliced } = useMemo(
    () => sliceFn(items, primaryCount),
    // Only recompute when the suggestions themselves change to avoid flashing
    // when the primaryCount toggles (e.g., empty vs filled input) without data changes.
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [items, sliceFn]
  );
  if (primaryCount === -1) {
    return {
      items,
      isSliced: false,
      isExpanded,
      controlProps
    };
  }
  return {
    items: isExpanded ? items : resultItems,
    isSliced,
    controlProps,
    isExpanded
  };
}
function defaultSlice(items, count) {
  if (items.length <= count + 1) {
    return { items, isSliced: false };
  }
  return { items: items.slice(0, count), isSliced: true };
}
function sliceFlattenedSuggestions(suggestions, count) {
  if (count === -1) {
    return { items: suggestions, isSliced: false };
  }
  let totalSuggestionCount = 0;
  for (const item of suggestions) {
    if (item.kind !== "group-header") {
      totalSuggestionCount++;
    }
  }
  if (totalSuggestionCount <= count + 1) {
    return { items: suggestions, isSliced: false };
  }
  let visibleCount = 0;
  const itemsShown = [];
  for (const item of suggestions) {
    if (item.kind === "group-header") {
      itemsShown.push(item);
      continue;
    }
    if (visibleCount >= count) {
      break;
    }
    itemsShown.push(item);
    visibleCount++;
  }
  if (itemsShown.length > 0 && itemsShown[itemsShown.length - 1].kind === "group-header") {
    itemsShown.pop();
  }
  return { items: itemsShown, isSliced: true };
}
export {
  sliceFlattenedSuggestions,
  useShowMore
};
