// @dynatrace/strato-components/esm/forms/select/hooks/use-select-options.js
import { useMemo, Children } from "react";
import { canReceiveDefaultTextFormat as _canReceiveDefaultTextFormat } from "../../../core/utils/can-receive-default-text-format.js";
import { extractBehaviorTrackingProps } from "../../../core/utils/extract-behavior-tracking-props.js";
import { isSlot } from "../../../core/utils/is-slot.js";
import {
  SelectGroupLabel,
  SelectGroup
} from "../SelectGroup.js";
import { _isSelectOptionElement } from "../SelectOption.js";
import { SelectPrefix } from "../SelectPrefix.js";
import { SelectSuffix } from "../SelectSuffix.js";
function _isOption(optionOrGroup) {
  if (optionOrGroup.value !== void 0) {
    return true;
  }
  return false;
}
function _isGroup(optionOrGroup) {
  if (Array.isArray(optionOrGroup.options)) {
    return true;
  }
  return false;
}
function extractOptionData(optionElement) {
  const optionSlots = {
    prefix: { Prefix: null },
    Label: null,
    suffix: { Suffix: null }
  };
  Children.forEach(optionElement.props.children, (child) => {
    if (isSlot(child, SelectPrefix)) {
      optionSlots.prefix.Prefix = child.props.children;
      optionSlots.prefix.componentProps = {
        className: child.props.className,
        style: child.props.style,
        "data-testid": child.props["data-testid"],
        ...extractBehaviorTrackingProps(child.props)
      };
    } else if (isSlot(child, SelectSuffix)) {
      optionSlots.suffix.Suffix = child.props.children;
      optionSlots.suffix.componentProps = {
        className: child.props.className,
        style: child.props.style,
        "data-testid": child.props["data-testid"],
        ...extractBehaviorTrackingProps(child.props)
      };
    } else {
      optionSlots.Label = optionSlots.Label ?? [];
      if (child !== null) {
        optionSlots.Label.push(child);
      }
    }
  });
  return {
    value: optionElement.props.value,
    prefix: {
      component: optionSlots.prefix.Prefix,
      componentProps: optionSlots.prefix.componentProps
    },
    label: optionSlots.Label,
    suffix: {
      component: optionSlots.suffix.Suffix,
      componentProps: optionSlots.suffix.componentProps
    },
    isDisabled: optionElement.props.disabled,
    textValue: optionElement.props.textValue,
    "aria-label": optionElement.props["aria-label"],
    className: optionElement.props.className,
    style: optionElement.props.style,
    "data-testid": optionElement.props["data-testid"],
    ...extractBehaviorTrackingProps(optionElement.props)
  };
}
function safeStringify(value) {
  try {
    if (typeof value === "string") {
      return value;
    }
    return JSON.stringify(value);
  } catch {
    return "";
  }
}
function reportNoStringifyableValueError(value, logger) {
  logger?.warn("Failed to stringify value.", value);
}
function getOptionLabel(option, fallbackLabel) {
  if (option.textValue) {
    return option.textValue;
  }
  const labelIsStringifyable = _canReceiveDefaultTextFormat(option.label);
  const stringifiedValue = safeStringify(option.value);
  if (labelIsStringifyable) {
    return `${option.label}`;
  }
  if (stringifiedValue) {
    return stringifiedValue;
  }
  return fallbackLabel;
}
function getOptionValue(option, fallbackValue, logger) {
  if (option.textValue) {
    return option.textValue;
  }
  const stringifiedValue = safeStringify(option.value);
  if (stringifiedValue) {
    return stringifiedValue;
  }
  const labelIsStringifyable = _canReceiveDefaultTextFormat(option.label);
  if (labelIsStringifyable) {
    return `${option.label}`;
  }
  reportNoStringifyableValueError(option.value, logger);
  return fallbackValue;
}
function extractGroupData(groupElement, groupIndex, logger) {
  const {
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId
  } = groupElement.props;
  const group = {
    label: void 0,
    options: [],
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    ...extractBehaviorTrackingProps(groupElement.props)
  };
  Children.forEach(groupElement?.props.children, (child, optionIndex) => {
    if (isSlot(child, SelectGroupLabel)) {
      group.label = child.props.children;
      group.labelProps = {
        className: child.props.className,
        style: child.props.style,
        "data-testid": child.props["data-testid"],
        ...extractBehaviorTrackingProps(child.props)
      };
    }
    if (_isSelectOptionElement(child)) {
      const option = extractOptionData(child);
      if (option.label) {
        const stringifiedLabel = getOptionLabel(
          option,
          `group-${groupIndex}-label-${optionIndex}`
        );
        option.label.toString = () => stringifiedLabel;
      }
      if (!option.textValue) {
        const stringifiedValue = getOptionValue(
          option,
          `group-${groupIndex}-label-${optionIndex}`,
          logger
        );
        option.textValue = stringifiedValue;
      }
      if (option !== null) {
        group.options.push(option);
      }
    }
  });
  return group;
}
function useSelectOptions(contentSlot, logger) {
  return useMemo(() => {
    const optionsOrGroups = [];
    let optionsHavePrefix = false;
    const allOptions = [];
    Children.forEach(contentSlot?.props.children, (child, childIndex) => {
      if (_isSelectOptionElement(child)) {
        const option = extractOptionData(child);
        if (option.label) {
          const stringifiedLabel = getOptionLabel(
            option,
            `label-${childIndex}`
          );
          option.label.toString = () => stringifiedLabel;
        }
        if (!option.textValue) {
          const stringifiedValue = getOptionValue(
            option,
            `value-${childIndex}`,
            logger
          );
          option.textValue = stringifiedValue;
        }
        if (option !== null) {
          optionsOrGroups.push(option);
          if (option.prefix.component !== null) {
            optionsHavePrefix = true;
          }
        }
        allOptions.push(option);
      }
      if (isSlot(child, SelectGroup)) {
        const group = extractGroupData(child, childIndex, logger);
        if (group !== null) {
          optionsOrGroups.push(group);
        }
        if (group.options.length > 2e4) {
          logger?.warn(
            `Group with label "${group.label}" has more than 20,000 options. This can lead to a slow and unresponsive UI.`
          );
        }
        for (const element of group.options) {
          allOptions.push(element);
        }
        if (optionsHavePrefix === false) {
          const nestedOptionsHavePrefix = group.options.some(
            (option) => option.prefix.component !== null
          );
          if (nestedOptionsHavePrefix) {
            optionsHavePrefix = true;
          }
        }
      }
    });
    return { optionsOrGroups, optionsHavePrefix, allOptions };
  }, [contentSlot, logger]);
}
export {
  _isGroup,
  _isOption,
  safeStringify,
  useSelectOptions
};
