// @dynatrace/strato-components/esm/forms/select/hooks/use-select-slots.js
import {
  Children,
  isValidElement,
  useMemo
} from "react";
import { isSlot } from "../../../core/utils/is-slot.js";
import { SelectContent } from "../SelectContent.js";
import { SelectCustomTrigger } from "../SelectCustomTrigger.js";
import { SelectDisplayValue } from "../SelectDisplayValue.js";
import {
  SelectEmptyState
} from "../SelectEmptyState.js";
import { SelectFilter } from "../SelectFilter.js";
import { SelectPrefix } from "../SelectPrefix.js";
import { SelectTrigger } from "../SelectTrigger.js";
function useSelectSlots(children) {
  return useMemo(() => {
    const slots = {
      Content: null,
      CustomTrigger: null,
      DisplayValue: null,
      Trigger: null,
      Prefix: null,
      Filter: null,
      EmptyState: null
    };
    Children.forEach(children, (child) => {
      if (isSlot(child, SelectContent)) {
        slots.Content = child;
        Children.forEach(child.props.children, (child2) => {
          if (isSlot(child2, SelectFilter)) {
            slots.Filter = child2;
            return;
          } else if (isSlot(child2, SelectEmptyState)) {
            slots.EmptyState = child2;
          }
        });
        return;
      } else if (isSlot(child, SelectTrigger)) {
        slots.Trigger = child;
        Children.forEach(child.props.children, (child2) => {
          if (isSlot(child2, SelectDisplayValue)) {
            Children.forEach(child2.props.children, (child3) => {
              if (isSlot(child3, SelectPrefix)) {
                slots.Prefix = child3;
                return;
              }
              slots.DisplayValue = slots.DisplayValue || [];
              slots.DisplayValue.push(child3);
            });
          }
        });
      } else if (isSlot(child, SelectCustomTrigger)) {
        const triggerValue = Children.toArray(child.props.children)[0];
        if (isValidElement(triggerValue)) {
          slots.CustomTrigger = triggerValue;
        }
      } else if (isSlot(child, SelectFilter)) {
        slots.Filter = child;
      }
    });
    return slots;
  }, [children]);
}
export {
  useSelectSlots
};
