// @dynatrace/strato-components/filters/filter-field/parser/character-codes.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var character_codes_exports = {};
__export(character_codes_exports, {
  characterCodes: () => characterCodes,
  durationUnitChars: () => durationUnitChars
});
module.exports = __toCommonJS(character_codes_exports);
const characterCodes = {
  EOF: -1,
  SPACE: 32,
  BANG: 33,
  DOUBLEQUOTE: 34,
  DOLLAR: 36,
  OPENING_BRACKET: 40,
  CLOSING_BRACKET: 41,
  ASTERISK: 42,
  COMMA: 44,
  DASH: 45,
  DOT: 46,
  EQUALS: 61,
  SQUARE_OPENING_BRACKET: 91,
  BACKSLASH: 92,
  TILDE: 126
};
const durationUnitChars = ["n", "s", "m", "h", "d"];
