// @dynatrace/strato-components/filters/filter-field/parser/getNumberParser.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var getNumberParser_exports = {};
__export(getNumberParser_exports, {
  NumberFormatParser: () => NumberFormatParser,
  getNumberParser: () => getNumberParser
});
module.exports = __toCommonJS(getNumberParser_exports);
class NumberFormatParser {
  formatter;
  getIndex;
  groupDelimiter;
  decimalDelimiter;
  numeralRegExp;
  groupRegExp;
  decimalRegExp;
  constructor(locale) {
    this.formatter = new Intl.NumberFormat(locale);
    const parts = this.formatter.formatToParts(12345.6789);
    const groupPart = parts.find((d) => d.type === "group");
    const decimalPart = parts.find((d) => d.type === "decimal");
    const numerals = [
      ...new Intl.NumberFormat(locale, { useGrouping: false }).format(
        9876543210
      )
    ].reverse();
    const indexMap = new Map(numerals.map((d, i) => [d, i]));
    this.getIndex = (d) => indexMap.get(d);
    this.numeralRegExp = new RegExp(`[${numerals.join("")}]`, "g");
    if (groupPart) {
      this.groupRegExp = new RegExp(`[${groupPart.value}]`, "g");
      this.groupDelimiter = groupPart.value;
    }
    if (decimalPart) {
      this.decimalRegExp = new RegExp(`[${decimalPart.value}]`);
      this.decimalDelimiter = decimalPart.value;
    }
  }
  parse(value) {
    const invalidDelimiter = value.match(
      new RegExp(
        `(\\${this.groupDelimiter}\\d{0,2}($|\\${this.groupDelimiter}))|(\\${this.decimalDelimiter}$)`
      )
    );
    if (invalidDelimiter !== null) {
      return NaN;
    }
    let parsed = value.trim();
    if (this.groupRegExp) {
      parsed = parsed.replace(this.groupRegExp, "");
    }
    if (this.decimalRegExp) {
      parsed = parsed.replace(this.decimalRegExp, ".");
    }
    if (this.numeralRegExp) {
      parsed = parsed.replace(
        this.numeralRegExp,
        (d) => this.getIndex(d)?.toString() ?? ""
      );
    }
    return parsed ? +parsed : NaN;
  }
}
let numberParserInstance;
function getNumberParser() {
  if (numberParserInstance === void 0) {
    numberParserInstance = new NumberFormatParser("en");
  }
  return numberParserInstance;
}
