// @dynatrace/strato-components/filters/filter-field/parser/keywords.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var keywords_exports = {};
__export(keywords_exports, {
  filterFieldKeywords: () => filterFieldKeywords
});
module.exports = __toCommonJS(keywords_exports);
const filterFieldKeywords = {
  "=": { keyword: ["=", "=*", "= *"], needsFilterKey: true },
  "~": { keyword: "~", needsFilterKey: true },
  "<": { keyword: ["<", "<="], needsFilterKey: true },
  ">": { keyword: [">", ">="], needsFilterKey: true },
  "!": { keyword: ["!=", "!~", "!=*", "!= *"], needsFilterKey: true },
  a: { keyword: "and", leading: [" ", "(", ")"], trailing: [" ", "(", ")"] },
  o: { keyword: "or", leading: [" ", "(", ")"], trailing: [" ", "(", ")"] },
  i: {
    keyword: "in",
    needsFilterKey: true,
    leading: [" "],
    trailing: [" (", "("]
  },
  n: { keyword: "not in", needsFilterKey: true, trailing: [" (", "("] },
  "*": { keyword: ["*~", "* ~"] },
  $: { keyword: ["$.", "$["], needsFilterKey: true }
};
