// @dynatrace/strato-components/filters/filter-field/parser/parser.terms.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var parser_terms_exports = {};
__export(parser_terms_exports, {
  Boolean: () => Boolean,
  ClosingBracket: () => ClosingBracket,
  ComparisonOperator: () => ComparisonOperator,
  Duration: () => Duration,
  ExistsOperator: () => ExistsOperator,
  Filter: () => Filter,
  FilterGroup: () => FilterGroup,
  FilterIncludesValue: () => FilterIncludesValue,
  FilterKey: () => FilterKey,
  FilterSimpleValue: () => FilterSimpleValue,
  FilterStatement: () => FilterStatement,
  FreeText: () => FreeText,
  Identifier: () => Identifier,
  InclusionOperator: () => InclusionOperator,
  IncompleteString: () => IncompleteString,
  InvalidComparisonOperator: () => InvalidComparisonOperator,
  InvalidIdentifier: () => InvalidIdentifier,
  InvalidInclusionValue: () => InvalidInclusionValue,
  InvalidPositionOperator: () => InvalidPositionOperator,
  JsonPath: () => JsonPath,
  JsonPathArray: () => JsonPathArray,
  JsonPathBase: () => JsonPathBase,
  JsonPathIdentifier: () => JsonPathIdentifier,
  JsonPathNameSelector: () => JsonPathNameSelector,
  JsonPathSeparator: () => JsonPathSeparator,
  JsonPathSquareOpeningBracket: () => JsonPathSquareOpeningBracket,
  List: () => List,
  ListValue: () => ListValue,
  LogicalOperator: () => LogicalOperator,
  Number: () => Number,
  OpeningBracket: () => OpeningBracket,
  PositionOperator: () => PositionOperator,
  SearchOperator: () => SearchOperator,
  String: () => String,
  Variable: () => Variable
});
module.exports = __toCommonJS(parser_terms_exports);
const Number = 1, Duration = 2, ExistsOperator = 3, LogicalOperator = 4, InclusionOperator = 5, SearchOperator = 6, PositionOperator = 7, InvalidPositionOperator = 8, String = 9, IncompleteString = 10, FreeText = 11, JsonPathIdentifier = 12, Filter = 13, FilterGroup = 14, OpeningBracket = 15, FilterStatement = 16, FilterKey = 17, Identifier = 18, InvalidIdentifier = 19, Variable = 20, JsonPath = 21, JsonPathBase = 22, JsonPathSeparator = 23, JsonPathArray = 24, JsonPathNameSelector = 25, JsonPathSquareOpeningBracket = 26, ComparisonOperator = 27, InvalidComparisonOperator = 28, FilterSimpleValue = 29, Boolean = 30, FilterIncludesValue = 31, List = 32, ClosingBracket = 33, ListValue = 34, InvalidInclusionValue = 35;
