// @dynatrace/strato-components/filters/filter-field/parser/parser-context.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var parser_context_exports = {};
__export(parser_context_exports, {
  trackLastToken: () => trackLastToken
});
module.exports = __toCommonJS(parser_context_exports);
var import_lr = require("@lezer/lr");
class LastAcceptedStatement {
  lastToken;
  pos;
  validExpression;
  constructor(lastToken, pos, validExpression) {
    this.lastToken = lastToken;
    this.pos = pos;
    this.validExpression = validExpression;
  }
}
const trackTokens = [
  "expression",
  "comparison",
  "ComparisonOperator",
  "InclusionOperator",
  "SearchOperator",
  "Identifier",
  "JsonPathBase",
  "PositionOperator",
  "ClosingBracket",
  '","'
];
const trackLastToken = new import_lr.ContextTracker({
  start: new LastAcceptedStatement(null, null),
  reduce(context, term, stack) {
    const termName = stack.parser.getName(term);
    if (trackTokens.includes(termName)) {
      return new LastAcceptedStatement(termName, stack.pos);
    }
    return context;
  },
  shift(context, term, stack) {
    const termName = stack.parser.getName(term);
    if (context.lastToken === "ComparisonOperator" && termName === "Identifier") {
      return new LastAcceptedStatement(context.lastToken, context.pos, true);
    }
    if (context.lastToken === "ComparisonOperator" && termName === "String") {
      return new LastAcceptedStatement(termName, stack.pos);
    }
    if (trackTokens.includes(termName)) {
      return new LastAcceptedStatement(termName, stack.pos);
    }
    return context;
  },
  strict: false
});
