// @dynatrace/strato-components/esm/forms/forms-core/time/date-time.js
import { TZDateMini } from "@date-fns/tz";
import { startOfDay } from "date-fns";
import { getTimezone } from "@dynatrace-sdk/user-preferences";
const GRAIL_MIN_DATE = "1677-09-21T00:12:43.145224192Z";
const GRAIL_MAX_DATE = "2262-04-11T23:47:16.854775807Z";
function getISOFormattedDate(date) {
  return `${valueTo4DigitString(date.getFullYear())}-${valueTo2DigitString(
    date.getMonth() + 1
  )}-${valueTo2DigitString(date.getDate())}`;
}
function getISOFormattedTime(date) {
  return `${valueTo2DigitString(date.getHours())}:${valueTo2DigitString(
    date.getMinutes()
  )}:${valueTo2DigitString(date.getSeconds())}`;
}
function getFormattedDateTime(date) {
  return `${getISOFormattedDate(date)} ${getISOFormattedTime(date)}`;
}
function getSecondFractions(searchString) {
  return searchString && searchString.includes(".") ? searchString.split(".")[1].replace("Z", "") : "";
}
function valueTo2DigitString(value) {
  return `${value}`.padStart(2, "0");
}
function valueTo4DigitString(value) {
  return `${value}`.padStart(4, "0");
}
function replaceSecondFractions(searchString, milliMicroNanoSeconds) {
  if (milliMicroNanoSeconds) {
    return `${searchString.replace(
      searchString.split(".")[1],
      milliMicroNanoSeconds
    )}Z`;
  }
  return searchString;
}
function getTimeFromISOString(isoDate) {
  return isoDate ? isoDate.split("T")[1].replace("Z", "") : "";
}
function applyInitialTime(newDate, originalDate, milliMicroNanoSeconds = getTimeFromISOString(
  startOfDay(TZDateMini.tz(getTimezone())).toISOString()
).split(".")[1]) {
  if (!originalDate) {
    return replaceSecondFractions(newDate, milliMicroNanoSeconds);
  }
  const configuredTimezone = getTimezone();
  const datePart = new TZDateMini(newDate, configuredTimezone);
  const timePart = new TZDateMini(originalDate, configuredTimezone);
  datePart.setMilliseconds(timePart.getMilliseconds());
  datePart.setUTCSeconds(timePart.getUTCSeconds());
  datePart.setMinutes(timePart.getMinutes());
  datePart.setHours(timePart.getHours());
  return datePart.toISOString();
}
export {
  GRAIL_MAX_DATE,
  GRAIL_MIN_DATE,
  applyInitialTime,
  getFormattedDateTime,
  getSecondFractions,
  replaceSecondFractions
};
