// @dynatrace/strato-components/esm/forms/forms-core/time/helpers.js
import { TZDateMini } from "@date-fns/tz";
import {
  addSeconds,
  constructFrom,
  endOfDay,
  isBefore,
  isSameDay,
  startOfDay
} from "date-fns";
import { parseTimeAsTimeValue } from "@dynatrace-sdk/units";
import {
  getRegionalFormat,
  getTimezone
} from "@dynatrace-sdk/user-preferences";
import {
  applyInitialTime,
  getSecondFractions,
  GRAIL_MAX_DATE,
  GRAIL_MIN_DATE
} from "./date-time.js";
import { removePrecision } from "./timeParser.js";
import { getStratoLogger } from "../../../core/sdk/sdk-logger.js";
import { isSimpleTimeOffset } from "../../../core/time/helpers.js";
import { isOutOfBounds } from "../../../core/time/validation.js";
import { isOutsideRange } from "../../../filters/timeframe-selector/utils/timeframe-selection.js";
function normalizeDates(context, ...dates) {
  const normalize = constructFrom.bind(
    null,
    context || dates.find((date) => typeof date === "object")
  );
  return dates.map(normalize);
}
function eachDayOfInterval(interval, options) {
  const [start, end] = normalizeDates(
    options?.in,
    interval.start,
    interval.end
  );
  let reversed = +start > +end;
  const endTime = reversed ? +start : +end;
  const date = reversed ? end : start;
  date.setHours(0, 0, 0, 0);
  let step = options?.step ?? 1;
  if (!step) {
    return [];
  }
  if (step < 0) {
    step = -step;
    reversed = !reversed;
  }
  const dates = [];
  while (+date <= endTime) {
    dates.push(constructFrom(start, date));
    date.setHours(date.getHours() + step * 24 + 1);
    date.setHours(0);
  }
  return reversed ? dates.reverse() : dates;
}
function sortTimeframeValues(value, min, max, precision) {
  if (value?.from?.absoluteDate && value?.to?.absoluteDate && isBefore(value.to.absoluteDate, value.from.absoluteDate)) {
    return {
      from: getTimeValueFromSimpleValue(
        value.to.value,
        value.from.value,
        "from",
        value.from.type === "expression",
        min,
        max,
        precision
      ),
      to: getTimeValueFromSimpleValue(
        value.from.value,
        value.to.value,
        "to",
        value.to.type === "expression",
        min,
        max,
        precision
      )
    };
  }
  return {
    from: value?.from,
    to: value?.to
  };
}
let hasLoggedSafeRegionalFormatError = false;
function resetRegionalFormatErrorFlag() {
  hasLoggedSafeRegionalFormatError = false;
}
function getSafeRegionalFormat() {
  const region = getRegionalFormat();
  const modifiedRegion = removeProblematicLanguagesFromRegion(region);
  try {
    new Intl.Locale(modifiedRegion);
    if (Intl.DisplayNames.supportedLocalesOf([modifiedRegion]).length === 0) {
      throw new Error(
        `Locale "${modifiedRegion}" is not supported by Intl.DisplayNames.`
      );
    }
    return modifiedRegion;
  } catch (error) {
    const logger = getStratoLogger("Date-Component");
    if (!hasLoggedSafeRegionalFormatError) {
      logger.warn(`Invalid regional format "${region}" encountered.`, error, {
        region,
        modifiedRegion
      });
      hasLoggedSafeRegionalFormatError = true;
    }
    return "en-US";
  }
}
const removeProblematicLanguagesFromRegion = (region) => {
  const problematicLanguages = [
    // Non-Arabic numerals.
    "ar",
    "as",
    "bgc",
    "bho",
    "bn",
    "ccp",
    "ckb",
    "dz",
    "fa",
    "ff",
    "ks",
    "lrc",
    "mni",
    "mr",
    "my",
    "mzn",
    "ne",
    "pa",
    "ps",
    "raj",
    "sa",
    "sat",
    "sd",
    "ur",
    "uz",
    // Problematic day period.
    "ak",
    "bem",
    "bo",
    "ee",
    "ha",
    "hi",
    "kok",
    "mai",
    "naq",
    "nus",
    "pt",
    "sq",
    "ti",
    "to",
    "yue",
    "zh",
    "tw",
    "bh",
    // Non gregorian calendars.
    "th"
  ];
  let safeRegionalFormat = region;
  if (!safeRegionalFormat) {
    return "en-US";
  }
  if (safeRegionalFormat.split("-")[1] === void 0) {
    return "en-" + safeRegionalFormat.split("-")[0];
  }
  if (problematicLanguages.find(
    (lang) => lang === safeRegionalFormat.split("-")[0].toLocaleLowerCase()
  )) {
    safeRegionalFormat = "en-" + safeRegionalFormat.split("-")[1];
  }
  if (safeRegionalFormat.split("-")[3]) {
    return safeRegionalFormat.split("-").slice(0, 2).join("-");
  }
  return safeRegionalFormat;
};
const getDayPeriodValues = () => {
  const dayPeriod = new Intl.DisplayNames(getSafeRegionalFormat(), {
    type: "dateTimeField"
  }).of("dayPeriod");
  if (!dayPeriod?.split("/")[1]) {
    return ["AM", "PM"];
  }
  return dayPeriod.split("/");
};
function timeValueToDate(timeValue, relativeDate) {
  if (!timeValue) {
    return null;
  }
  const parsedString = parseTimeAsTimeValue(timeValue?.value, relativeDate);
  return parsedString?.absoluteDate ? new TZDateMini(parsedString.absoluteDate, getTimezone()) : null;
}
const getTimeValueFromSimpleValue = (candidate, timeOffset, position, isExpression, min, max, precision) => {
  if (!candidate) {
    return null;
  }
  const stringDate = typeof candidate === "string" ? candidate : candidate.toISOString();
  const timezone = getTimezone();
  const date = new TZDateMini(stringDate, timezone);
  const minDate = new TZDateMini(
    removePrecision(min?.absoluteDate ?? "", precision),
    timezone
  );
  if (min && position !== "to" && isOutOfBounds(startOfDay(date), min) && isSameDay(minDate, date)) {
    if (min.type === "expression") {
      return min;
    }
    const offsetByPrecision = { milliseconds: 0, seconds: 1, minutes: 60 };
    const offsetSeconds = (
      // Only add the offset when the min has different precision than the calendar.
      minDate.getTime() === new TZDateMini(min.absoluteDate, timezone).getTime() ? 0 : offsetByPrecision[precision] ?? 0
    );
    const parsedMin = parseTimeAsTimeValue(
      addSeconds(minDate, offsetSeconds).toISOString(),
      void 0,
      precision
    );
    if (parsedMin && isSameDay(minDate, date)) {
      return parsedMin;
    }
  }
  if (max && position !== "from" && isOutOfBounds(endOfDay(date), void 0, max) && isSameDay(new TZDateMini(max.absoluteDate, timezone), date)) {
    return max;
  }
  const parsedDate = parseTimeAsTimeValue(stringDate, void 0, precision);
  if (!parsedDate) {
    return null;
  }
  const isRelativeExpression = isExpression || parsedDate.type === "expression";
  const resolvedDate = precision === "day" ? parsedDate.absoluteDate : _getPreciseIsoString(
    parsedDate.absoluteDate,
    timeOffset,
    position,
    isRelativeExpression
  );
  return isRelativeExpression ? { ...parsedDate, absoluteDate: resolvedDate, type: "expression" } : { ...parsedDate, absoluteDate: resolvedDate, value: resolvedDate };
};
function _getPreciseIsoString(isoString, timeOffset, position, isExpression) {
  if (!timeOffset) {
    return isoString;
  }
  const offsetDate = isSimpleTimeOffset(timeOffset) ? timeOffset : timeOffset[position ?? "from"];
  const date = new TZDateMini(isoString, getTimezone());
  if (isExpression) {
    return date.toISOString();
  }
  switch (offsetDate) {
    case "startOfDay":
      return startOfDay(date).toISOString();
    case "endOfDay":
      return endOfDay(date).toISOString();
    default: {
      if (isoString === offsetDate) {
        return isoString;
      }
      const parsedOffsetDate = parseTimeAsTimeValue(offsetDate);
      if (parsedOffsetDate?.type === "expression") {
        return parsedOffsetDate.absoluteDate;
      }
      const precision = getSecondFractions(offsetDate);
      return applyInitialTime(isoString, offsetDate, precision);
    }
  }
}
const convertDateOnlyToDateTime = (isoString, type) => {
  if (type === "date") {
    const newDate = new TZDateMini(isoString ?? "", getTimezone());
    if (!isNaN(newDate.getTime())) {
      newDate.setMinutes(newDate.getMinutes() + newDate.getTimezoneOffset());
      return newDate.toISOString();
    }
  }
  return isoString ?? "";
};
function toCalendarValue(value, precision) {
  if (!value || value.absoluteDate === "") {
    return null;
  }
  if (precision === "day") {
    const converted = convertDateOnlyToDateTime(value.absoluteDate, "date");
    if (converted) {
      return { absoluteDate: converted, value: converted, type: "iso8601" };
    }
  }
  return value;
}
function filterOutGrailLimit(value, limit) {
  const isOutsideGrailLimit = isOutsideRange(
    value,
    GRAIL_MIN_DATE,
    GRAIL_MAX_DATE
  );
  return (limit === GRAIL_MIN_DATE || limit === GRAIL_MAX_DATE) && !isOutsideGrailLimit ? void 0 : limit;
}
export {
  convertDateOnlyToDateTime,
  eachDayOfInterval,
  filterOutGrailLimit,
  getDayPeriodValues,
  getSafeRegionalFormat,
  getTimeValueFromSimpleValue,
  resetRegionalFormatErrorFlag,
  sortTimeframeValues,
  timeValueToDate,
  toCalendarValue
};
