// @dynatrace/strato-components/charts/pie/components/PieChartToolbar/components/DownloadDataButton.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DownloadDataButton_exports = {};
__export(DownloadDataButton_exports, {
  DownloadDataButton: () => DownloadDataButton
});
module.exports = __toCommonJS(DownloadDataButton_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_ChartToolbarRenderer = require("../../../../core/components/toolbar/ChartToolbarRenderer.js");
var import_useA11yToolbarTabIndex = require("../../../../core/hooks/useA11yToolbarTabIndex.js");
var import_PieChartRawDataContext = require("../../../context/PieChartRawDataContext.js");
var import_messages = require("../../../messages.js");
var import_build_and_download_csv = require("../../../utils/build-and-download-csv.js");
const DownloadDataButton = () => {
  const downloadDataButtonRef = (0, import_react.useRef)(null);
  const tabIndex = (0, import_useA11yToolbarTabIndex.useA11yToolbarTabIndex)("download");
  const intl = (0, import_react_intl.useIntl)();
  const rawData = (0, import_PieChartRawDataContext.usePieChartRawData)();
  const handleDownloadDataKeyboard = (event) => {
    if (event.key === "Enter") {
      (0, import_build_and_download_csv.buildAndDownloadCsv)(rawData);
    }
  };
  const handleDownloadDataToolbar = () => {
    (0, import_build_and_download_csv.buildAndDownloadCsv)(rawData);
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ChartToolbarRenderer.ChartToolbarRenderer.Control,
    {
      ref: downloadDataButtonRef,
      prefixIcon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.DownloadIcon, {}),
      tooltipText: intl.formatMessage(import_messages.PIE_CHART_MESSAGES.downloadAsCsv),
      testId: "chart-toolbar-download",
      onKeyDown: handleDownloadDataKeyboard,
      onClick: handleDownloadDataToolbar,
      tabIndex
    }
  );
};
DownloadDataButton["displayName"] = "DownloadDataButton";
