// @dynatrace/strato-components/esm/charts/xy-chart/XYChartRoot.js
import { jsx, jsxs } from "react/jsx-runtime";
import { forwardRef, useRef } from "react";
import { XYChartLegend } from "./components/legend/XYChartLegend.js";
import {
  DEFAULT_CHART_HEIGHT,
  DEFAULT_LOADING_STATE
} from "./constants/defaults.js";
import { XYChartAxesConfigContext } from "./context/axes/XYChartAxesConfig.context.js";
import { XYChartAnnotationsConfig } from "./context/XYChartAnnotationsConfig.context.js";
import { XYChartColorPaletteContext } from "./context/XYChartColorPalette.context.js";
import { XYChartLegendHiddenContext } from "./context/XYChartLegendHidden.context.js";
import { XYChartThresholdsContext } from "./context/XYChartThresholds.context.js";
import { XYChartToolbarContext } from "./context/XYChartToolbar.context.js";
import { XYChartTruncationModeContext } from "./context/XYChartTruncationMode.context.js";
import { useXYChartConfig } from "./hooks/useXYChartConfig.js";
import { XYChartAnnotationsDataProvider } from "./providers/XYChartAnnotationsDataProvider.js";
import { XYChartAnnotationsShownIndicatorsProvider } from "./providers/XYChartAnnotationsShownIndicatorsProvider.js";
import { XYChartColorScalesProvider } from "./providers/XYChartColorScalesProvider.js";
import { XYChartChartDataProvider } from "./providers/XYChartDataProvider.js";
import { XYChartDownloadProvider } from "./providers/XYChartDownloadProvider.js";
import { XYChartIdProvider } from "./providers/XYChartIdProvider.js";
import { XYChartStoreDomainSync } from "./providers/XYChartStoreDomainSync.js";
import { XYChartStoreProvider } from "./providers/XYChartStoreProvider.js";
import { XYChartVisibleAxesProvider } from "./providers/XYChartVisibleAxesProvider.js";
import { Annotations } from "./slots/Annotations/Annotations.js";
import { Marker } from "./slots/Annotations/Marker.js";
import { AnnotationsTooltip } from "./slots/Annotations/Tooltip.js";
import { Track } from "./slots/Annotations/Track.js";
import { AreaSeries } from "./slots/AreaSeries.js";
import { BarSeries } from "./slots/BarSeries.js";
import { CustomTooltip } from "./slots/CustomTooltip.js";
import { DotSeries } from "./slots/DotSeries.js";
import { Legend } from "./slots/Legend.js";
import { LineSeries } from "./slots/LineSeries.js";
import { RectSeries } from "./slots/RectSeries.js";
import { Select } from "./slots/Select.js";
import { ThresholdIndicator } from "./slots/ThresholdIndicator.js";
import { Toolbar } from "./slots/Toolbar.js";
import { XYChartTooltipSlot as Tooltip } from "./slots/Tooltip.js";
import { XAxis } from "./slots/XAxis.js";
import { YAxis } from "./slots/YAxis.js";
import { isEmptyData } from "./utils/is-empty-data.js";
import { iterateSlots } from "./utils/iterate-slots.js";
import { buildInternalConfig } from "./utils/xy-chart-configuration.js";
import { XYChart as XYChartInternal } from "./XYChart.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useMergeRefs } from "../../core/hooks/useMergeRefs.js";
import { Intent } from "../../core/slots/Intent/Intent.js";
import { ChartLayout } from "../core/components/chart-layout/ChartLayout.js";
import { useAutoLegendRefresh } from "../core/components/chart-layout/hooks/useAutoLegendRefresh.js";
import { getChartStatesTemplates } from "../core/components/chart-states/utils/get-chart-states-template.js";
import { OverlayProvider } from "../core/components/overlay-tooltip/contexts/OverlayContext.js";
import { TooltipProvider } from "../core/components/overlay-tooltip/contexts/TooltipContext.js";
import { ColorRuleContext } from "../core/context/ColorRuleContext.js";
import { EmptyStateSlot } from "../core/slots/chart-states/EmptyStateSlot.js";
import { ErrorStateSlot } from "../core/slots/chart-states/ErrorStateSlot.js";
import { ColorRule } from "../core/slots/ColorRule.js";
import { coerceSizeValue } from "../core/utils/coerceSizeValue.js";
const COMPONENT_NAME_BEHAVIOURAL_TRACKING = "XYChart";
const XYChartRoot = /* @__PURE__ */ forwardRef((props, ref) => {
  const {
    data,
    children,
    height,
    width,
    loading = DEFAULT_LOADING_STATE,
    style,
    className,
    __downloadDataCallback: downloadDataCallback,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow
  } = props;
  const [behaviorTrackingProps] = useBehavioralTrackingProps(props);
  const slots = iterateSlots(props.children);
  const config = useXYChartConfig(props, slots);
  const { emptyState, errorState } = getChartStatesTemplates(props.children);
  const containerRef = useRef(null);
  const rootRef = useMergeRefs([
    ref,
    containerRef
  ]);
  const {
    hidden: legendHidden,
    // eslint-disable-next-line @typescript-eslint/no-deprecated
    ratio,
    initialRatio,
    onRatioChange,
    position: legendPosition
  } = config.legend;
  const legendRatio = initialRatio ?? ratio;
  const chartHeight = coerceSizeValue(height ?? DEFAULT_CHART_HEIGHT);
  const internalConfig = buildInternalConfig(config, slots);
  const chartLayoutRef = useAutoLegendRefresh(data);
  return /* @__PURE__ */ jsx(XYChartIdProvider, { children: /* @__PURE__ */ jsx(XYChartThresholdsContext.Provider, { value: internalConfig.thresholds, children: /* @__PURE__ */ jsx(XYChartAxesConfigContext.Provider, { value: internalConfig.axes, children: /* @__PURE__ */ jsx(XYChartColorPaletteContext.Provider, { value: config.colorPalette, children: /* @__PURE__ */ jsx(
    XYChartDownloadProvider,
    {
      downloadDataCallback,
      children: /* @__PURE__ */ jsx(XYChartToolbarContext.Provider, { value: config.toolbar, children: /* @__PURE__ */ jsx(
        XYChartTruncationModeContext.Provider,
        {
          value: { truncationMode: config.truncationMode },
          children: /* @__PURE__ */ jsx(
            XYChartLegendHiddenContext.Provider,
            {
              value: Boolean(legendHidden),
              children: /* @__PURE__ */ jsx(
                XYChartAnnotationsConfig.Provider,
                {
                  value: config.annotations,
                  children: /* @__PURE__ */ jsx(ColorRuleContext.Provider, { value: config.colorRules, children: /* @__PURE__ */ jsx(XYChartStoreProvider, { children: /* @__PURE__ */ jsx(
                    XYChartChartDataProvider,
                    {
                      data,
                      seriesVariantsProps: internalConfig.series,
                      children: /* @__PURE__ */ jsx(XYChartStoreDomainSync, { children: /* @__PURE__ */ jsx(XYChartAnnotationsDataProvider, { children: /* @__PURE__ */ jsx(XYChartVisibleAxesProvider, { children: /* @__PURE__ */ jsx(XYChartColorScalesProvider, { children: /* @__PURE__ */ jsx(XYChartAnnotationsShownIndicatorsProvider, { children: /* @__PURE__ */ jsx(OverlayProvider, { children: /* @__PURE__ */ jsx(TooltipProvider, { children: /* @__PURE__ */ jsxs(
                        ChartLayout,
                        {
                          ref: chartLayoutRef,
                          chartHeight,
                          chartWidth: width,
                          className,
                          emptyState,
                          errorState,
                          isEmpty: isEmptyData(
                            children,
                            data
                          ),
                          showLoader: loading,
                          style,
                          "data-dtrum-mask": dataDtrumMask,
                          "data-dtrum-allow": dataDtrumAllow,
                          ...behaviorTrackingProps,
                          "data-dt-component": COMPONENT_NAME_BEHAVIOURAL_TRACKING,
                          children: [
                            /* @__PURE__ */ jsx(ChartLayout.Graph, { children: /* @__PURE__ */ jsx(
                              XYChartInternal,
                              {
                                config,
                                ref: rootRef
                              }
                            ) }),
                            !legendHidden && /* @__PURE__ */ jsx(
                              ChartLayout.Legend,
                              {
                                position: legendPosition,
                                initialRatio: legendRatio,
                                onResize: onRatioChange,
                                children: /* @__PURE__ */ jsx(XYChartLegend, {})
                              }
                            )
                          ]
                        }
                      ) }) }) }) }) }) }) })
                    }
                  ) }) })
                }
              )
            }
          )
        }
      ) })
    }
  ) }) }) }) });
});
const XYChart = Object.assign(XYChartRoot, {
  Annotations,
  RectSeries,
  DotSeries,
  LineSeries,
  BarSeries,
  AreaSeries,
  XAxis,
  YAxis,
  Legend,
  Toolbar,
  /** @deprecated Use `XYChart.Tooltip` instead. */
  // eslint-disable-next-line @typescript-eslint/no-deprecated
  CustomTooltip,
  Tooltip,
  Intent,
  Select,
  ThresholdIndicator,
  EmptyState: EmptyStateSlot,
  ErrorState: ErrorStateSlot,
  ColorRule
});
XYChart.displayName = "XYChart";
const XYChartAnnotations = {
  Marker,
  Track,
  Tooltip: AnnotationsTooltip
};
export {
  XYChart,
  XYChartAnnotations,
  XYChartRoot
};
