// @dynatrace/strato-components/overlays/modal/BaseModal.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var BaseModal_exports = {};
__export(BaseModal_exports, {
  BaseModal: () => BaseModal
});
module.exports = __toCommonJS(BaseModal_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_framer_motion = require("framer-motion");
var import_react = require("react");
var import_animations_js = __toESM(require("@dynatrace/strato-design-tokens/animations-js"));
var styles = __toESM(require("./Modal.sty.js"));
var import_FocusScope = require("../../core/components/focus-scope/FocusScope.js");
var import_OverlayContainer = require("../../core/components/OverlayContainer.js");
var import_useOverlayStacking = require("../../core/hooks/useOverlayStacking.js");
var import_OverlayStackingProvider = require("../../core/providers/OverlayStackingProvider.js");
var import_framer_motion_easings = require("../../core/utils/framer-motion-easings.js");
var import_useDialog = require("../hooks/useDialog.js");
var import_useOverlay = require("../hooks/useOverlay.js");
const BaseModal = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    show = false,
    size = "medium",
    children,
    onDismiss,
    dismissible = true,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    className: consumerClassName,
    style: consumerStyle,
    "data-dt-component": dtComponentName,
    ...remainingProps
  } = props;
  const ref = (0, import_react.useRef)(null);
  const { overlayProps } = (0, import_useOverlay.useOverlay)(
    {
      ...props,
      onClose: () => {
        onDismiss?.();
      },
      isOpen: show,
      isKeyboardDismissDisabled: !dismissible
    },
    ref
  );
  const { dialogProps, titleProps } = (0, import_useDialog.useDialog)(remainingProps);
  const overlayContainerProps = (0, import_useOverlayStacking.useOverlayStacking)(void 0, "overlay");
  const reducedMotion = (0, import_framer_motion.useReducedMotion)();
  const animationVariants = (0, import_react.useMemo)(() => {
    const duration = reducedMotion ? 0 : import_animations_js.default.FadeIn.Duration.Fast;
    return {
      // When animating AnimatePresence where child components
      // will also need to transform, they can share the same enter/exit variant states.
      // The keys will need to be exactly the same, and the child will not need to
      // define `initial`, `animate` or `exit` variants on the component.
      backdrop: {
        enter: {
          // For the opacity we need to use the default value here. Framer motion would complain
          // about too complex values, when transitioning from two custom property values within
          // an rgba value apparently.
          opacity: 1,
          transition: {
            duration,
            ease: import_framer_motion_easings.fadeInEasing
          }
        },
        exit: {
          opacity: 0,
          transition: {
            duration,
            ease: import_framer_motion_easings.fadeOutEasing
          }
        }
      },
      modal: {
        enter: {
          y: 0,
          scale: 1,
          transition: {
            duration,
            ease: import_framer_motion_easings.fadeInEasing
          }
        },
        exit: {
          y: "5px",
          scale: 0.98,
          transition: {
            duration,
            ease: import_framer_motion_easings.fadeInEasing
          }
        }
      }
    };
  }, [reducedMotion]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_OverlayContainer.OverlayContainer,
    {
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...overlayContainerProps,
      "data-dt-component": dtComponentName,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_framer_motion.AnimatePresence, { children: show && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_framer_motion.motion.div,
        {
          className: styles.backdrop,
          variants: animationVariants.backdrop,
          initial: "exit",
          animate: "enter",
          exit: "exit",
          ref: forwardedRef,
          onPointerDown: (event) => {
            if (event.target === event.currentTarget && dismissible) {
              onDismiss?.();
            }
          },
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_FocusScope.FocusScope, { contain: true, autoFocus: true, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_framer_motion.motion.div,
            {
              ...overlayProps,
              ...dialogProps,
              ...titleProps,
              className: (0, import_clsx.default)(
                styles.base,
                styles.sizes[size],
                consumerClassName
              ),
              style: consumerStyle,
              ref,
              variants: animationVariants.modal,
              ...remainingProps,
              children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayStackingProvider.OverlayStackingProvider, { children })
            }
          ) })
        }
      ) })
    }
  );
});
BaseModal.displayName = "BaseModal";
