// @dynatrace/strato-components/overlays/modal/Modal.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Modal_exports = {};
__export(Modal_exports, {
  Modal: () => Modal
});
module.exports = __toCommonJS(Modal_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_BaseModal = require("./BaseModal.js");
var styles = __toESM(require("./Modal.sty.js"));
var import_Button = require("../../buttons/button/Button.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useId = require("../../core/hooks/useId.js");
var import_can_receive_default_text_format = require("../../core/utils/can-receive-default-text-format.js");
var import_Divider = require("../../layouts/divider/Divider.js");
var import_safe_sprinkles = require("../../styles/safe-sprinkles.js");
var import_Heading = require("../../typography/heading/Heading.js");
const COMPONENT_NAME = "Modal";
function getVisualHeadingLevelForSize(size) {
  switch (size) {
    case "small":
      return 5;
    case "medium":
    case "large":
      return 4;
    default:
      return 4;
  }
}
const Modal = /* @__PURE__ */ (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const {
      show = false,
      title,
      footer,
      size = "medium",
      children,
      onDismiss,
      dismissible = true,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      className: consumerClassName,
      style: consumerStyle,
      ...remainingProps
    } = props;
    const visualHeadingSize = getVisualHeadingLevelForSize(size);
    const intl = (0, import_react_intl.useIntl)();
    const headerId = (0, import_useId.useId)();
    const closeButtonLabel = intl.formatMessage({
      defaultMessage: "Close modal",
      id: "aB3uCeB71q6UURBy",
      description: "Aria-label for close modal button."
    });
    const hasCustomTitle = !(0, import_can_receive_default_text_format.canReceiveDefaultTextFormat)(title);
    const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      import_BaseModal.BaseModal,
      {
        "aria-labelledby": headerId,
        show,
        size,
        onDismiss,
        dismissible,
        ref: forwardedRef,
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        className: consumerClassName,
        style: consumerStyle,
        ...restProps,
        ...behavioralTrackingProps,
        "data-dt-component": COMPONENT_NAME,
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("header", { className: styles.header, children: [
            hasCustomTitle ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { style: { display: "contents" }, id: headerId, children: title }) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_Heading.Heading,
              {
                id: headerId,
                level: visualHeadingSize,
                as: "h2",
                style: {
                  wordBreak: "normal",
                  overflowWrap: "anywhere"
                },
                children: title
              }
            ),
            dismissible && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_Button.Button,
              {
                style: { alignSelf: "start" },
                onClick: onDismiss,
                "aria-label": closeButtonLabel,
                variant: "default",
                "data-dt-name": "CloseX",
                children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.XmarkIcon, {}) })
              }
            )
          ] }),
          (children || footer) && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "div",
            {
              className: (0, import_safe_sprinkles.globalSprinkles)({
                paddingBottom: 0,
                paddingTop: 0,
                paddingLeft: 24,
                paddingRight: 24
              }),
              children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Divider.Divider, {})
            }
          ),
          children && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "div",
            {
              className: (0, import_safe_sprinkles.globalSprinkles)({
                paddingTop: 12,
                paddingRight: 24,
                paddingLeft: 24,
                paddingBottom: footer !== void 0 ? 12 : 24
              }),
              style: { overflowY: "auto" },
              children
            }
          ),
          footer && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "footer",
            {
              className: (0, import_safe_sprinkles.globalSprinkles)({
                display: "flex",
                alignItems: "flex-end",
                justifyContent: "flex-end",
                paddingBottom: 24,
                paddingTop: 12,
                paddingLeft: 24,
                paddingRight: 24
              }),
              children: footer
            }
          )
        ]
      }
    );
  }
);
Modal.displayName = COMPONENT_NAME;
