// @dynatrace/strato-components/esm/charts/pie/hooks/useSliceHandlers.js
import { useTooltip } from "../../core/components/overlay-tooltip/contexts/TooltipContext.js";
import { useTooltipActions } from "../../core/components/overlay-tooltip/useTooltipActions.js";
import { useEventErrorHandler } from "../../core/hooks/useEventErrorHandler.js";
import { getTooltipContent } from "../utils/get-tooltip-content.js";
const buildTooltipData = (sliceId, data, handlerCatch, formatter) => {
  const tooltipContent = getTooltipContent(
    sliceId,
    data,
    handlerCatch,
    formatter
  );
  return {
    metadata: {
      sliceId: data.id
    },
    geometry: "pie",
    series: [
      {
        name: tooltipContent.name,
        value: tooltipContent.absoluteValue,
        color: tooltipContent.color,
        relativeValue: tooltipContent.relativeValue,
        relativeValueUnit: "%"
      }
    ]
  };
};
function useSliceHandlers(sliceId, data, formatter) {
  const handlerCatch = useEventErrorHandler();
  const tooltipState = useTooltip();
  const { hover, pin, unpin, reset } = useTooltipActions();
  const onPointerMoveHandler = (event) => {
    if (tooltipState.getState().pinned || sliceId === "slice-thinner-slices") {
      return;
    }
    const tooltipData = buildTooltipData(
      sliceId,
      data,
      handlerCatch,
      formatter
    );
    hover(event, tooltipData);
  };
  const onMouseClickHandler = (event) => {
    event.stopPropagation();
    if (sliceId === "slice-thinner-slices") {
      reset();
    } else {
      const tooltipData = buildTooltipData(
        sliceId,
        data,
        handlerCatch,
        formatter
      );
      if (tooltipState.getState().pinned) {
        unpin();
        hover(event, tooltipData);
      } else {
        pin(event, tooltipData);
      }
    }
  };
  return {
    onPointerMoveHandler,
    onMouseClickHandler
  };
}
export {
  useSliceHandlers
};
