// @dynatrace/strato-components/charts/core/components/single-value-renderer/components/Label.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Label_exports = {};
__export(Label_exports, {
  Label: () => Label
});
module.exports = __toCommonJS(Label_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_Label_css = require("./Label.sty.js");
var import_useId = require("../../../../../core/hooks/useId.js");
var import_useMergeRefs = require("../../../../../core/hooks/useMergeRefs.js");
var import_Flex = require("../../../../../layouts/flex/Flex.js");
var import_useFontSizeForContainer = require("../../../hooks/useFontSizeForContainer.js");
var import_useThrottledResizeObserver = require("../../../hooks/useThrottledResizeObserver.js");
var import_constants = require("../constants.js");
var import_single_value_context = require("../contexts/single-value.context.js");
var import_use_global_font_size = require("../hooks/use-global-font-size.js");
const Label = ({ color }) => {
  const { label, alignment } = (0, import_react.useContext)(import_single_value_context.SingleValueContext);
  const ref = (0, import_react.useRef)(null);
  const { refCallback, width = 0, height = 0 } = (0, import_useThrottledResizeObserver.useThrottledResizeObserver)();
  const mergedRef = (0, import_useMergeRefs.useMergeRefs)([refCallback, ref]);
  const preferredFontSize = (0, import_useFontSizeForContainer.useFontSizeForContainer)({
    container: { width, height: height * import_constants.FONT_HEIGHT_RATIO },
    text: label ?? "",
    element: ref.current
  });
  const id = (0, import_useId.useId)();
  const { updateLabelFontSize } = (0, import_use_global_font_size.useGlobalFontSize)();
  (0, import_react.useLayoutEffect)(() => {
    if (height && width && label) {
      updateLabelFontSize?.(id, preferredFontSize);
    }
  }, [height, width, label, preferredFontSize, updateLabelFontSize, id]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Flex.Flex,
    {
      className: import_Label_css.labelAreaStaticStyles,
      style: {
        [import_constants.LABEL_FONT_CSS_TOKEN]: preferredFontSize,
        color
      },
      ref: mergedRef,
      justifyContent: alignment,
      alignItems: "center",
      children: label !== void 0 && /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { "data-testid": "single-value-chart-label", children: label })
    }
  );
};
Label["displayName"] = "SingleValue.Label";
