// @dynatrace/strato-components/charts/core/components/single-value-renderer/components/SingleValueToolbar.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SingleValueToolbar_exports = {};
__export(SingleValueToolbar_exports, {
  SingleValueToolbar: () => SingleValueToolbar
});
module.exports = __toCommonJS(SingleValueToolbar_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_SingleValueToolbar_css = require("./SingleValueToolbar.sty.js");
var import_constants = require("../../../constants.js");
var import_zoom = require("../../../utils/zoom.js");
var import_ChartToolbarRenderer = require("../../toolbar/ChartToolbarRenderer.js");
var import_constants2 = require("../constants.js");
const SingleValueToolbar = (props) => {
  const {
    parentHeight,
    hidden,
    intents,
    intentsAsyncTemplate,
    placement = "top-right",
    draggable
  } = props;
  const hasEnoughHeightForToolbar = parentHeight > import_constants2.MINIMUM_HEIGHT_FOR_TOOLBAR;
  const isToolbarEnabled = hasEnoughHeightForToolbar && !hidden && !!(intents?.length ?? intentsAsyncTemplate);
  if (!isToolbarEnabled) {
    return null;
  }
  const overlayClassName = parentHeight > import_constants2.MINIMUM_HEIGHT_FOR_TOOLBAR_WITH_PADDING ? void 0 : import_SingleValueToolbar_css.toolbarOverlayPadding;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ChartToolbarRenderer.ChartToolbarRenderer,
    {
      visible: false,
      placement,
      initialA11yState: (0, import_zoom.getA11yToolbarConfig)({
        intent: !!intents
      }),
      menuDisplayMode: "never",
      intents,
      overlayClassName,
      visibleWhenMouseInBounds: true,
      intentsAsyncTemplate,
      zIndex: import_constants.ChartOverlaysElevation.toolbar,
      draggable,
      removeDragHandle: true
    }
  );
};
SingleValueToolbar["displayName"] = "SingleValueToolbar";
