// @dynatrace/strato-components/content/code-snippet/utils/highlighting-utils.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var highlighting_utils_exports = {};
__export(highlighting_utils_exports, {
  splitCodeForHighlighting: () => splitCodeForHighlighting
});
module.exports = __toCommonJS(highlighting_utils_exports);
const DEFAULT_MAX_TOKENIZATION_LENGTH = 1e5;
function countNewlines(str) {
  let count = 0;
  for (let i = 0; i < str.length; i++) {
    if (str.codePointAt(i) === 10) {
      count++;
    }
  }
  return count;
}
function buildLineNumbers(startLine, lineCount) {
  const numbers = new Array(lineCount);
  for (let i = 0; i < lineCount; i++) {
    numbers[i] = String(startLine + i);
  }
  return numbers.join("\n");
}
function splitCodeForHighlighting(code, maxLength = DEFAULT_MAX_TOKENIZATION_LENGTH) {
  if (code.length <= maxLength) {
    return {
      highlightCode: code,
      noHighlightInline: "",
      noHighlightCode: "",
      noHighlightLineNumbers: ""
    };
  }
  const highlight = code.substring(0, maxLength);
  const remaining = code.substring(maxLength);
  const firstNewline = remaining.indexOf("\n");
  const inlinePart = firstNewline < 0 ? remaining : remaining.substring(0, firstNewline);
  const blockPart = firstNewline < 0 ? "" : remaining.substring(firstNewline + 1);
  if (blockPart.length === 0) {
    return {
      highlightCode: highlight,
      noHighlightInline: inlinePart,
      noHighlightCode: "",
      noHighlightLineNumbers: ""
    };
  }
  const highlightedLineCount = countNewlines(highlight) + 1;
  const startLine = highlightedLineCount + 1;
  const blockLineCount = countNewlines(blockPart) + 1;
  return {
    highlightCode: highlight,
    noHighlightInline: inlinePart,
    noHighlightCode: blockPart,
    noHighlightLineNumbers: buildLineNumbers(startLine, blockLineCount)
  };
}
