// @dynatrace/strato-components/content/skeleton/SkeletonText.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SkeletonText_exports = {};
__export(SkeletonText_exports, {
  SkeletonText: () => SkeletonText
});
module.exports = __toCommonJS(SkeletonText_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_Skeleton_css = require("./Skeleton.sty.js");
var import_utils = require("./utils.js");
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useId = require("../../core/hooks/useId.js");
var import_useSafeForwardProps = require("../../core/hooks/useSafeForwardProps.js");
const COMPONENT_NAME = "SkeletonText";
const lineHeight = 12;
const linePadding = 8;
const baseYOffset = linePadding * 0.5;
const lineCapSize = lineHeight * 0.5;
const lineStep = lineHeight + linePadding;
const getSkeletonLinesMargin = (line, lines) => {
  if (lines === 1) {
    return 0;
  }
  return (line + 2) % 3 * lineHeight;
};
const SkeletonText = (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const {
      width = "100%",
      lines = 3,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
    const ariaLabelingProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(restProps);
    const forwardProps = (0, import_useSafeForwardProps.useSafeForwardProps)(restProps, ariaLabelingProps);
    const maskId = (0, import_useId.useId)();
    const gradientId = (0, import_useId.useId)();
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "div",
      {
        ref: forwardedRef,
        className: consumerClassName,
        style: {
          width,
          height: `${lineStep * lines}px`,
          ...consumerStyle
        },
        "data-testid": dataTestId,
        ...ariaLabelingProps,
        ...behavioralTrackingProps,
        ...forwardProps,
        "data-dt-component": COMPONENT_NAME,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("svg", { className: import_Skeleton_css.skeletonSvgCSS, children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)("defs", { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_utils.SkeletonGradient, { id: gradientId }) }),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)("mask", { id: maskId, children: new Array(lines).fill(null).map((_, lineIndex) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "rect",
            {
              y: baseYOffset + lineStep * lineIndex,
              height: lineHeight,
              rx: lineCapSize,
              fill: "white",
              style: {
                // we also substract 1px, to make sure the line cap is fully within the svg, and isn't cut
                width: `calc(100% - 1px - ${getSkeletonLinesMargin(
                  lineIndex,
                  lines
                )}px)`
              }
            },
            `${lines}-${lineIndex}`
          )) }),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)("g", { mask: `url(#${maskId})`, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_utils.SkeletonDrawer, { gradientId }) })
        ] })
      }
    );
  }
);
SkeletonText.displayName = COMPONENT_NAME;
