// @dynatrace/strato-components/content/skeleton/utils.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var utils_exports = {};
__export(utils_exports, {
  SkeletonBackgroundFill: () => SkeletonBackgroundFill,
  SkeletonDrawer: () => SkeletonDrawer,
  SkeletonGradient: () => SkeletonGradient
});
module.exports = __toCommonJS(utils_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_timings_js = __toESM(require("@dynatrace/strato-design-tokens/timings-js"));
var import_Skeleton_css = require("./Skeleton.sty.js");
const gradientWidth = 400;
const SkeletonGradient = (props) => /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
  "linearGradient",
  {
    className: import_Skeleton_css.skeletonSvgGradientCSS,
    id: props.id,
    x1: "0",
    x2: "1",
    children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("stop", { offset: "0%", stopOpacity: "0.0" }),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("stop", { offset: "50%", stopOpacity: "1.0" }),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("stop", { offset: "100%", stopOpacity: "0.0" })
    ]
  }
);
const SkeletonBackgroundFill = () => /* @__PURE__ */ (0, import_jsx_runtime.jsx)("rect", { className: import_Skeleton_css.skeletonSvgBackgroundFillCSS });
const SkeletonDrawer = (props) => /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(SkeletonBackgroundFill, {}),
  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "rect",
    {
      className: import_Skeleton_css.skeletonSvgGradientRectCSS,
      width: `${gradientWidth}px`,
      height: "100%",
      fill: `url(#${props.gradientId})`,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        "animate",
        {
          attributeName: "x",
          from: `-${gradientWidth}px`,
          to: "100%",
          dur: `${import_timings_js.default.Slowest.Gentle}s`,
          repeatCount: "indefinite",
          keyTimes: "0; 1",
          calcMode: "spline",
          keySplines: "0.4, 0, 0.2, 1"
        }
      )
    }
  )
] });
