// @dynatrace/strato-components/content/progress/ProgressCircle.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ProgressCircle_exports = {};
__export(ProgressCircle_exports, {
  ProgressCircle: () => ProgressCircle
});
module.exports = __toCommonJS(ProgressCircle_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_useProgressAriaProps = require("./hooks/useProgressAriaProps.js");
var import_ProgressCircle_css = require("./ProgressCircle.sty.js");
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useId = require("../../core/hooks/useId.js");
var import_useSafeForwardProps = require("../../core/hooks/useSafeForwardProps.js");
var import_Container = require("../../layouts/container/Container.js");
var import_container_css = require("../../styles/container.sty.js");
var import_textStyle_css = require("../../styles/textStyle.sty.js");
const COMPONENT_NAME = "ProgressCircle";
const progressCircleSizes = {
  small: 20,
  large: 40
};
const progressCircleThickness = 5;
const ProgressCircle = (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const {
      value: originalValue,
      min = 0,
      max = 100,
      size = "large",
      color,
      children,
      id: propId,
      "aria-valuetext": ariaValuetext,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
    const { value, isIndeterminate, progressAriaProps } = (0, import_useProgressAriaProps.useProgressAriaProps)(
      originalValue,
      min,
      max,
      ariaValuetext
    );
    const containerColoring = (0, import_Container.useContainerColoring)();
    const derivedVariant = containerColoring.variant === "accent" ? "accent" : "default";
    const fallbackColor = containerColoring.default ? "primary" : containerColoring.color;
    const derivedColor = color ?? fallbackColor;
    const ariaLabelingProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(restProps);
    const forwardProps = (0, import_useSafeForwardProps.useSafeForwardProps)(restProps, ariaLabelingProps);
    const generatedId = (0, import_useId.useId)();
    const id = propId ?? generatedId;
    const circleStyles = {};
    if (!isIndeterminate && value !== void 0) {
      const circumference = 2 * Math.PI * ((progressCircleSizes[size] - progressCircleThickness) / 2);
      circleStyles["strokeDasharray"] = circumference.toFixed(3);
      const dashOffset = Number(
        ((max - value) / max * circumference).toFixed(3)
      );
      circleStyles["strokeDashoffset"] = `${dashOffset >= 0 ? dashOffset : 0}px`;
    }
    if (!children && !ariaLabelingProps["aria-label"] && !ariaLabelingProps["aria-labelledby"]) {
      // eslint-disable-next-line no-console
      console.warn(
        "If no label is provided please make sure to provide an aria label on your progressCircle element."
      );
    }
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      "div",
      {
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        ref: forwardedRef,
        className: (0, import_clsx.default)(
          consumerClassName,
          (0, import_container_css.containerColorsCSS)({
            color: derivedColor,
            variant: derivedVariant
          }),
          (0, import_ProgressCircle_css.progressCircleContainerCSS)({
            size,
            color: derivedColor,
            variant: derivedVariant
          })
        ),
        style: consumerStyle,
        ...behavioralTrackingProps,
        ...forwardProps,
        "data-dt-component": COMPONENT_NAME,
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "span",
            {
              id,
              className: (0, import_clsx.default)(import_ProgressCircle_css.progressCircleRootCSS),
              style: {
                height: `${progressCircleSizes[size]}px`,
                width: `${progressCircleSizes[size]}px`
              },
              ...ariaLabelingProps,
              ...progressAriaProps,
              children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
                "svg",
                {
                  className: import_ProgressCircle_css.progressCircleSvgCss,
                  viewBox: `0 0 ${progressCircleSizes[size]} ${progressCircleSizes[size]}`,
                  width: "100%",
                  height: "100%",
                  children: [
                    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                      "circle",
                      {
                        fill: "none",
                        className: (0, import_clsx.default)(import_ProgressCircle_css.progressCircleBackgroundCSS),
                        cx: progressCircleSizes[size] / 2,
                        cy: progressCircleSizes[size] / 2,
                        r: (progressCircleSizes[size] - progressCircleThickness) / 2,
                        strokeWidth: progressCircleThickness
                      }
                    ),
                    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                      "circle",
                      {
                        style: circleStyles,
                        className: (0, import_ProgressCircle_css.progressCircleCSS)({ indeterminate: isIndeterminate }),
                        fill: "none",
                        pathLength: isIndeterminate ? 360 : circleStyles["strokeDasharray"],
                        cx: progressCircleSizes[size] / 2,
                        cy: progressCircleSizes[size] / 2,
                        r: (progressCircleSizes[size] - progressCircleThickness) / 2,
                        strokeWidth: progressCircleThickness
                      }
                    )
                  ]
                }
              )
            }
          ),
          children && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "label",
            {
              className: (0, import_clsx.default)(
                (0, import_textStyle_css.textStyle)({ textStyle: size === "small" ? "small" : "base" })
              ),
              htmlFor: id,
              children
            }
          )
        ]
      }
    );
  }
);
ProgressCircle.displayName = COMPONENT_NAME;
