// @dynatrace/strato-components/content/progress/ProgressBar.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ProgressBar_exports = {};
__export(ProgressBar_exports, {
  ProgressBar: () => ProgressBar
});
module.exports = __toCommonJS(ProgressBar_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_is = require("react-is");
var import_SharedProgressBarPropsContext = require("./contexts/SharedProgressBarPropsContext.js");
var import_useProgressAriaProps = require("./hooks/useProgressAriaProps.js");
var import_ProgressBar_css = require("./ProgressBar.sty.js");
var import_ProgressBarIcon = require("./ProgressBarIcon.js");
var import_ProgressBarLabel = require("./ProgressBarLabel.js");
var import_ProgressBarValue = require("./ProgressBarValue.js");
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useId = require("../../core/hooks/useId.js");
var import_useSafeForwardProps = require("../../core/hooks/useSafeForwardProps.js");
var import_Container = require("../../layouts/container/Container.js");
var import_Flex = require("../../layouts/flex/Flex.js");
var import_container_css = require("../../styles/container.sty.js");
const COMPONENT_NAME = "ProgressBar";
function getProgressBarSlots(children) {
  const slots = {
    value: null,
    label: null,
    icon: null
  };
  import_react.Children.forEach(children, (child) => {
    if ((0, import_react_is.isElement)(child)) {
      switch (child.type) {
        case import_ProgressBarLabel.ProgressBarLabel:
          slots.label = child;
          break;
        case import_ProgressBarIcon.ProgressBarIcon:
          slots.icon = child;
          break;
        case import_ProgressBarValue.ProgressBarValue:
        default:
          slots.value = child;
          break;
      }
    } else {
      if (!slots.value) {
        slots.value = child;
      }
    }
  });
  return slots;
}
const _ProgressBar = (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const {
      value: originalValue,
      max = 100,
      min = 0,
      density = "default",
      color,
      children,
      id: propId,
      "aria-valuetext": ariaValuetext,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...remainingProps
    } = props;
    const containerColoring = (0, import_Container.useContainerColoring)();
    const fallbackColor = containerColoring.default ? "primary" : containerColoring.color;
    const derivedColor = color ?? fallbackColor;
    const onAccent = containerColoring.variant === "accent";
    const slots = getProgressBarSlots(children);
    const { value, isIndeterminate, progressAriaProps } = (0, import_useProgressAriaProps.useProgressAriaProps)(
      originalValue,
      min,
      max,
      ariaValuetext
    );
    const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
    const ariaLabelingProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(restProps);
    const forwardProps = (0, import_useSafeForwardProps.useSafeForwardProps)(restProps, ariaLabelingProps);
    const generatedId = (0, import_useId.useId)();
    const id = propId ?? generatedId;
    if (!slots.label && !ariaLabelingProps["aria-label"] && !ariaLabelingProps["aria-labelledby"]) {
      // eslint-disable-next-line no-console
      console.warn(
        "If no label is provided please make sure to provide an aria label on your progressCircle element."
      );
    }
    const progressStyles = {};
    if (!isIndeterminate && value !== void 0) {
      const percentage = (value - min) / (max - min);
      progressStyles["width"] = `${Math.round(percentage * 100)}%`;
    }
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_SharedProgressBarPropsContext.SharedProgressBarPropsContext.Provider, { value: { id, density }, children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      import_Flex.Flex,
      {
        ref: forwardedRef,
        flexDirection: "column",
        gap: 6,
        width: "100%",
        className: (0, import_clsx.default)(
          (0, import_container_css.containerColorsCSS)({
            color: derivedColor,
            variant: containerColoring.variant
          }),
          (0, import_ProgressBar_css.progressRoot)({
            onAccent,
            color: derivedColor
          }),
          consumerClassName
        ),
        style: consumerStyle,
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        ...behavioralTrackingProps,
        ...forwardProps,
        "data-dt-component": COMPONENT_NAME,
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { justifyContent: "space-between", gap: 8, children: [
            density === "condensed" ? /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
              import_Flex.Flex,
              {
                justifyContent: "space-between",
                alignItems: "flex-end",
                flex: 1,
                gap: 16,
                minWidth: 0,
                children: [
                  slots.label,
                  slots.value
                ]
              }
            ) : slots.label,
            slots.icon
          ] }),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "div",
            {
              className: (0, import_ProgressBar_css.progressCSS)({
                size: density
              }),
              id,
              ...progressAriaProps,
              ...ariaLabelingProps,
              children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                "div",
                {
                  className: (0, import_clsx.default)(
                    (0, import_ProgressBar_css.progressBarFillCSS)({
                      indeterminate: isIndeterminate
                    })
                  ),
                  style: progressStyles
                }
              )
            }
          ),
          density === "default" && slots.value
        ]
      }
    ) });
  }
);
_ProgressBar.displayName = COMPONENT_NAME;
const ProgressBar = Object.assign(_ProgressBar, {
  Value: import_ProgressBarValue.ProgressBarValue,
  Label: import_ProgressBarLabel.ProgressBarLabel,
  Icon: import_ProgressBarIcon.ProgressBarIcon
});
