// @dynatrace/strato-components/content/progress/ProgressBarValue.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ProgressBarValue_exports = {};
__export(ProgressBarValue_exports, {
  ProgressBarValue: () => ProgressBarValue
});
module.exports = __toCommonJS(ProgressBarValue_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_useProgressBarPropsContext = require("./hooks/useProgressBarPropsContext.js");
var import_ProgressBarValue_css = require("./ProgressBarValue.sty.js");
var import_can_receive_default_text_format = require("../../core/utils/can-receive-default-text-format.js");
var import_textStyle_css = require("../../styles/textStyle.sty.js");
const ProgressBarValue = (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    children,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    ...remainingProps
  } = props;
  const intl = (0, import_react_intl.useIntl)();
  const progressBarDefaultLabel = intl.formatMessage({
    defaultMessage: "Loading",
    id: "3J6qHQEUwDxvZuIl",
    description: "Default label for the progress-bar."
  });
  const { density } = (0, import_useProgressBarPropsContext.useSharedProgressBarPropsContext)();
  const childIsText = (0, import_can_receive_default_text_format.canReceiveDefaultTextFormat)(children);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      "data-testid": dataTestId,
      ref: forwardedRef,
      className: (0, import_clsx.default)(
        consumerClassName,
        (0, import_ProgressBarValue_css.value)({ size: density, childIsText }),
        (0, import_textStyle_css.textStyle)({ textStyle: "small" })
      ),
      style: consumerStyle,
      ...remainingProps,
      children: children || progressBarDefaultLabel
    }
  );
});
ProgressBarValue.displayName = "ProgressBar.Value";
