// @dynatrace/strato-components/charts/meter-bar/shared/components/meter-bar-layout/MeterBarLayout.sty.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MeterBarLayout_css_exports = {};
__export(MeterBarLayout_css_exports, {
  chartContainer: () => chartContainer,
  container: () => container,
  containerWrapper: () => containerWrapper,
  labelContainer: () => labelContainer,
  legendContainer: () => legendContainer,
  maxContainer: () => maxContainer,
  minContainer: () => minContainer,
  toolbarAnimatedContainer: () => toolbarAnimatedContainer,
  valueAnimatedContainer: () => valueAnimatedContainer,
  valueContainer: () => valueContainer
});
module.exports = __toCommonJS(MeterBarLayout_css_exports);
var import_MeterBarLayout_css_ts_vanilla = require("./MeterBarLayout.css");
var import_createRuntimeFn = require("@vanilla-extract/recipes/createRuntimeFn");
var chartContainer = "_1m3udw19-3-6-0";
var container = (0, import_createRuntimeFn.createRuntimeFn)({ defaultClassName: "_1m3udw1a-3-6-0", variantClassNames: { autoSize: { true: "_1m3udw1b-3-6-0", false: "_1m3udw1c-3-6-0" }, showLoading: { true: "_1m3udw1d-3-6-0" } }, defaultVariants: {}, compoundVariants: [] });
var containerWrapper = "_1m3udw1e-3-6-0";
var labelContainer = "_1m3udw18-3-6-0";
var legendContainer = "_1m3udw17-3-6-0";
var maxContainer = "_1m3udw16-3-6-0";
var minContainer = "_1m3udw15-3-6-0";
var toolbarAnimatedContainer = "_1m3udw14-3-6-0";
var valueAnimatedContainer = "_1m3udw13-3-6-0";
var valueContainer = (0, import_createRuntimeFn.createRuntimeFn)({ defaultClassName: "_1m3udw10-3-6-0", variantClassNames: { inline: { true: "_1m3udw11-3-6-0", false: "_1m3udw12-3-6-0" } }, defaultVariants: {}, compoundVariants: [] });
