// @dynatrace/strato-components/charts/meter-bar/shared/components/meter-bar-layout/MeterBarLayout.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MeterBarLayout_exports = {};
__export(MeterBarLayout_exports, {
  MeterBarLayout: () => MeterBarLayout,
  MeterBarLayoutRoot: () => MeterBarLayoutRoot
});
module.exports = __toCommonJS(MeterBarLayout_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_MeterBarLayoutTopRightSlot = require("./components/MeterBarLayoutTopRightSlot.js");
var import_useComputedSegmentSize = require("./hooks/useComputedSegmentSize.js");
var import_MeterBarLayout_css = require("./MeterBarLayout.sty.js");
var import_Chart = require("./slots/Chart.js");
var import_Icon = require("./slots/Icon/Icon.js");
var import_Label = require("./slots/Label/Label.js");
var import_Legend = require("./slots/Legend.js");
var import_Max = require("./slots/Max/Max.js");
var import_Min = require("./slots/Min/Min.js");
var import_Providers = require("./slots/Providers.js");
var import_Value = require("./slots/Value/Value.js");
var import_extract_slots = require("./utils/extract-slots.js");
var import_useBehavioralTrackingProps = require("../../../../../core/hooks/useBehavioralTrackingProps.js");
var import_useMergeRefs = require("../../../../../core/hooks/useMergeRefs.js");
var import_Flex = require("../../../../../layouts/flex/Flex.js");
var import_Grid = require("../../../../../layouts/grid/Grid.js");
var import_LoaderOverlay = require("../../../../core/components/loader-overlay/LoaderOverlay.js");
var import_useTooltipImperativeHandle = require("../../../../core/components/overlay-tooltip/hooks/useTooltipImperativeHandle.js");
var import_coerceSizeValue = require("../../../../core/utils/coerceSizeValue.js");
var import_constants = require("../../constants.js");
var import_MeterBarChartLoadingContext = require("../../context/MeterBarChartLoadingContext.js");
var import_useMeterBarChartToolbar = require("../../hooks/useMeterBarChartToolbar.js");
var import_styles_css = require("../../styles.sty.js");
var import_MeterBarSizeProviders = require("../meter-bar-providers/MeterBarSizeProviders.js");
const MeterBarLayoutRoot = (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    width,
    height,
    size,
    children,
    className,
    style,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "data-testid": dataTestId,
    "data-dt-component": dataDtComponent,
    "aria-label": ariaLabel,
    userRef,
    consumerConfig = {}
  } = props;
  const containerRef = (0, import_react.useRef)(null);
  const { label, min, max, value, chart, icon, legend, providers } = (0, import_extract_slots.extractSlots)(children);
  const [behavioralTrackingProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(props);
  const loading = (0, import_MeterBarChartLoadingContext.useMeterBarChartLoading)();
  const { isToolbarEnabled, modifyHoverState, toolbarRef } = (0, import_useMeterBarChartToolbar.useMeterBarChartToolbar)();
  const tooltipActions = (0, import_useTooltipImperativeHandle.useTooltipImperativeHandle)();
  const ref = (0, import_useMergeRefs.useMergeRefs)([containerRef, forwardedRef]);
  (0, import_react.useImperativeHandle)(userRef, () => {
    return {
      element: containerRef.current,
      getConfig: () => JSON.stringify(consumerConfig),
      ...tooltipActions
    };
  });
  const hasAutoSize = size === import_constants.METER_BAR_SIZE.Auto;
  const gridTemplateRows = hasAutoSize ? "auto 1fr auto auto" : "auto";
  const {
    segmentSize: { height: segmentHeight, width: segmentWidth },
    labelRef,
    minLabelRef,
    maxLabelRef,
    plotAreaRef,
    legendRef
  } = (0, import_useComputedSegmentSize.useComputedSegmentSize)();
  const chartWidth = (0, import_coerceSizeValue.coerceSizeValue)(width) ?? import_constants.DEFAULT_WIDTH;
  const chartHeight = (0, import_coerceSizeValue.coerceSizeValue)(height) ?? import_constants.DEFAULT_HEIGHT;
  const ExternalProviders = providers ? providers.Component : import_react.Fragment;
  const externalProvidersProps = providers ? providers.props : {};
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      ...behavioralTrackingProps,
      className: (0, import_clsx.default)(className, import_MeterBarLayout_css.containerWrapper),
      style: {
        width: chartWidth,
        height: chartHeight,
        ...style
      },
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "data-dt-component": dataDtComponent,
      ref,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        "div",
        {
          className: (0, import_MeterBarLayout_css.container)({ showLoading: loading, autoSize: hasAutoSize }),
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_Grid.Grid,
            {
              className: (0, import_MeterBarLayout_css.container)({ autoSize: hasAutoSize }),
              gridTemplateAreas: "\n            'label label label value'\n            'chart chart chart chart'\n            'min min max max'\n            'legend legend legend legend'",
              gridTemplateColumns: "25% 25% 25% 25%",
              gridTemplateRows,
              rowGap: 0,
              columnGap: 0,
              "aria-label": ariaLabel,
              role: "img",
              onMouseEnter: () => modifyHoverState(true),
              onMouseLeave: () => modifyHoverState(false),
              onFocus: () => modifyHoverState(true),
              onBlur: () => modifyHoverState(false),
              children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_MeterBarSizeProviders.MeterBarSizeProviders,
                {
                  size,
                  chartWidth: containerRef.current?.clientWidth ?? 0,
                  chartHeight: containerRef.current?.clientHeight ?? 0,
                  segmentWidth,
                  segmentHeight,
                  "data-dtrum-allow": dataDtrumAllow,
                  "data-dtrum-mask": dataDtrumMask,
                  children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(ExternalProviders, { ...externalProvidersProps, children: [
                    (label || icon) && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
                      import_Flex.Flex,
                      {
                        ref: labelRef,
                        alignItems: "center",
                        alignSelf: "center",
                        className: import_MeterBarLayout_css.labelContainer,
                        children: [
                          icon,
                          label
                        ]
                      }
                    ),
                    (isToolbarEnabled || value) && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                      import_MeterBarLayoutTopRightSlot.MeterBarLayoutTopRightSlot,
                      {
                        value,
                        isToolbarEnabled,
                        ref: toolbarRef
                      }
                    ),
                    /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: import_MeterBarLayout_css.chartContainer, ref: plotAreaRef, children: chart }),
                    min && /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { ref: minLabelRef, className: import_MeterBarLayout_css.minContainer, children: min }),
                    max && /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { ref: maxLabelRef, className: import_MeterBarLayout_css.maxContainer, children: max }),
                    legend && /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { ref: legendRef, className: import_MeterBarLayout_css.legendContainer, children: legend }),
                    loading ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                      import_LoaderOverlay.LoaderOverlay,
                      {
                        gridArea: "1/1/2/2",
                        consumerStyle: import_styles_css.consumerLoadingStyle
                      }
                    ) : null
                  ] })
                }
              )
            }
          )
        }
      )
    }
  );
});
MeterBarLayoutRoot.displayName = "MeterBarLayoutRoot";
const MeterBarLayout = Object.assign(MeterBarLayoutRoot, {
  Label: import_Label.Label,
  Min: import_Min.Min,
  Max: import_Max.Max,
  Chart: import_Chart.Chart,
  Value: import_Value.Value,
  Icon: import_Icon.Icon,
  Providers: import_Providers.Providers,
  Legend: import_Legend.Legend
});
