// @dynatrace/strato-components/content/keyboard-shortcut/KeyboardShortcut.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var KeyboardShortcut_exports = {};
__export(KeyboardShortcut_exports, {
  KeyboardShortcut: () => KeyboardShortcut,
  createKeyArrayFromString: () => createKeyArrayFromString
});
module.exports = __toCommonJS(KeyboardShortcut_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_KeyboardShortcut_css = require("./KeyboardShortcut.sty.js");
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useSafeForwardProps = require("../../core/hooks/useSafeForwardProps.js");
var import_Container = require("../../layouts/container/Container.js");
const COMPONENT_NAME = "KeyboardShortcut";
const keyMappings = {
  meta: "Cmd",
  control: "Ctrl",
  escape: "Esc",
  altgraph: "Option",
  delete: "Del",
  arrowup: "\u2191",
  arrowdown: "\u2193",
  arrowright: "\u2192",
  arrowleft: "\u2190",
  plus: "+",
  minus: "-",
  multiply: "*",
  divide: "/"
};
function createKeyArrayFromString(keys) {
  return keys.split(" ")[0].split("+").filter((key) => key !== "").map(
    (key) => keyMappings[key.toLowerCase()] ?? `${key.charAt(0).toUpperCase()}${key.slice(1)}`
  );
}
const KeyboardShortcut = (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const {
      keys,
      variant = "default",
      disabled = false,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...remainingProps
    } = props;
    const keyArray = (0, import_react.useMemo)(() => createKeyArrayFromString(keys), [keys]);
    const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
    const ariaLabelingProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(restProps);
    const forwardProps = (0, import_useSafeForwardProps.useSafeForwardProps)(restProps, ariaLabelingProps);
    const containerColoring = (0, import_Container.useContainerColoring)();
    const onAccent = containerColoring.variant === "accent";
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "kbd",
      {
        ref: forwardedRef,
        "aria-disabled": disabled,
        className: (0, import_clsx.default)(import_KeyboardShortcut_css.keyboardShortcutKeyWrapperCSS, consumerClassName),
        style: consumerStyle,
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        ...ariaLabelingProps,
        ...behavioralTrackingProps,
        ...forwardProps,
        "data-dt-component": COMPONENT_NAME,
        children: keyArray.map((key, index) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          "kbd",
          {
            className: (0, import_KeyboardShortcut_css.keyboardShortcutKeyCSS)({ variant, onAccent, disabled }),
            children: key
          },
          `${key}-${index}`
        ))
      }
    );
  }
);
KeyboardShortcut.displayName = COMPONENT_NAME;
