// @dynatrace/strato-components/esm/charts/core/components/legend/categorical/utils/legend-grid.utils.js
import { charsFontWidth12pixelsSize500Map } from "../../../text-ellipsis/chars-font-width-map.js";
import { calculateWidth } from "../../../text-ellipsis/truncate-string.js";
import {
  ITEM_HEIGHT,
  LEGEND_BUTTON_WIDTH,
  MIN_ITEM_WIDTH,
  VIRTUALIZED_GRID_THRESHOLD
} from "../constants.js";
const getMinMaxSeriesWidth = (items) => {
  const { shortestSeries, longestSeries } = items.reduce(
    (acc, current) => {
      const labelWidth = calculateWidth(current.name, charsFontWidth12pixelsSize500Map) + LEGEND_BUTTON_WIDTH;
      return {
        shortestSeries: Math.min(acc.shortestSeries, labelWidth),
        longestSeries: Math.max(acc.longestSeries, labelWidth)
      };
    },
    {
      shortestSeries: Infinity,
      longestSeries: -Infinity
    }
  );
  return {
    shortestSeries,
    longestSeries
  };
};
const buildGridSettings = (items, width, height) => {
  const { shortestSeries, longestSeries } = getMinMaxSeriesWidth(items);
  const maxRowsCanFitTheContainer = Math.floor(height / ITEM_HEIGHT.value);
  const maxColumnsCanFitTheContainer = Math.max(
    Math.floor(width / MIN_ITEM_WIDTH),
    1
  );
  const minItemWidth = Math.min(
    Math.max(shortestSeries, MIN_ITEM_WIDTH),
    width
  );
  const maxItemWidth = Math.min(Math.max(longestSeries, MIN_ITEM_WIDTH), width);
  const maxItemsPerRow = Math.floor(width / maxItemWidth);
  const maxNrOfSeriesTruncated = maxItemsPerRow * maxRowsCanFitTheContainer;
  const canFitAllItemsWithoutScrollbar = maxNrOfSeriesTruncated >= items.length;
  return {
    maxRows: maxRowsCanFitTheContainer,
    maxColumns: maxColumnsCanFitTheContainer,
    minItemWidth,
    maxItemWidth,
    canFitAllItemsWithoutScrollbar,
    hasVirtual: !canFitAllItemsWithoutScrollbar && items.length > VIRTUALIZED_GRID_THRESHOLD,
    longestSeries
  };
};
export {
  buildGridSettings,
  getMinMaxSeriesWidth
};
