// @dynatrace/strato-geo/map/utils/build-geo-tooltip-state.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var build_geo_tooltip_state_exports = {};
__export(build_geo_tooltip_state_exports, {
  buildGeoTooltipState: () => buildGeoTooltipState
});
module.exports = __toCommonJS(build_geo_tooltip_state_exports);
const buildGeoTooltipState = ({
  absoluteX,
  absoluteY,
  relativeX,
  relativeY,
  pinned,
  geometry,
  metadata
}) => ({
  absoluteX,
  absoluteY,
  relativeX,
  relativeY,
  percentX: 0,
  percentY: 0,
  inBounds: true,
  pinned,
  activeSeries: null,
  data: {
    geometry,
    series: [],
    metadata
  }
});
