// @dynatrace/strato-geo/map/utils/iterate-config-slots.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var iterate_config_slots_exports = {};
__export(iterate_config_slots_exports, {
  iterateConfigSlots: () => iterateConfigSlots
});
module.exports = __toCommonJS(iterate_config_slots_exports);
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_charts = require("@dynatrace/strato-components/charts");
var import_constants = require("../constants.js");
var import_is_slot = require("./is-slot.js");
var import_replace_infinite_values_in_ranges = require("./replace-infinite-values-in-ranges.js");
var import_BaseLayer = require("../slots/BaseLayer.js");
var import_CategoricalLegend = require("../slots/CategoricalLegend.js");
var import_ChartInteractions = require("../slots/ChartInteractions.js");
var import_DownloadCSV = require("../slots/DownloadCSV.js");
var import_SequentialLegend = require("../slots/SequentialLegend.js");
var import_ThresholdLegend = require("../slots/ThresholdLegend.js");
var import_Toolbar = require("../slots/Toolbar.js");
var import_Zoom = require("../slots/Zoom.js");
const iterateConfigSlots = (children, legendDomain) => {
  let slots = {
    toolbar: void 0,
    interactions: void 0,
    legend: void 0,
    baseLayer: import_constants.DEFAULT_BASE_LAYER_RULES
  };
  let slotMapToolbarConfig = {};
  import_react.Children.forEach(children, (child) => {
    if ((0, import_is_slot.isSlot)(child, import_ChartInteractions.ChartInteractions)) {
      const { children: children2, ...previousInteractions } = child.props;
      const interactions = iterateChartInteractionsSlots(children2);
      slots = {
        ...slots,
        interactions: {
          ...previousInteractions,
          ...interactions
        }
      };
    } else if ((0, import_is_slot.isSlot)(child, import_charts.ChartToolbar)) {
      const { children: children2, ...toolbar } = child.props;
      const childrenToolbar = iterateChartToolbarSlots(children2);
      slots = {
        ...slots,
        toolbar: {
          ...toolbar,
          ...childrenToolbar
        }
      };
    } else if ((0, import_is_slot.isSlot)(child, import_SequentialLegend.SequentialLegend)) {
      slots = {
        ...slots,
        legend: {
          type: "sequential",
          colorPalette: import_constants.DEFAULT_SEQUENTIAL_LEGEND_COLOR_PALETTE,
          position: import_constants.DEFAULT_MAP_POSITION,
          ratio: import_constants.DEFAULT_MAP_LEGEND_RATIO,
          ...child.props,
          min: child.props.min ?? legendDomain[0],
          max: child.props.max ?? legendDomain[1]
        }
      };
    } else if ((0, import_is_slot.isSlot)(child, import_CategoricalLegend.CategoricalLegend)) {
      slots = {
        ...slots,
        legend: {
          type: "categorical",
          position: import_constants.DEFAULT_MAP_POSITION,
          ...child.props
        }
      };
    } else if ((0, import_is_slot.isSlot)(child, import_ThresholdLegend.ThresholdLegend)) {
      const getDomainValue = (index, legendDomainValue) => Number.isFinite(legendDomainValue) ? legendDomainValue : child.props.ranges?.at(index)?.to ?? legendDomainValue;
      const ranges = (0, import_replace_infinite_values_in_ranges.replaceInfiniteValuesInRanges)(child.props.ranges, [
        getDomainValue(0, legendDomain[0]),
        getDomainValue(-1, legendDomain[1])
      ]);
      slots = {
        ...slots,
        legend: {
          type: "threshold",
          position: import_constants.DEFAULT_MAP_POSITION,
          ratio: import_constants.DEFAULT_MAP_LEGEND_RATIO,
          ...child.props,
          ranges
        }
      };
    } else if ((0, import_is_slot.isSlot)(child, import_Toolbar.Toolbar)) {
      slotMapToolbarConfig = {
        ...slotMapToolbarConfig,
        ...(0, import_Toolbar.sanitizeProps)(child.props)
      };
    } else if ((0, import_is_slot.isSlot)(child, import_DownloadCSV.DownloadCSV)) {
      slotMapToolbarConfig = {
        ...slotMapToolbarConfig,
        downloadCSV: { enabled: true }
      };
    } else if ((0, import_is_slot.isSlot)(child, import_Zoom.Zoom)) {
      slotMapToolbarConfig = {
        ...slotMapToolbarConfig,
        zoom: { enabled: child.props.disabled !== true }
      };
    } else if ((0, import_is_slot.isSlot)(child, import_BaseLayer.BaseLayer)) {
      slots = {
        ...slots,
        baseLayer: {
          ...slots.baseLayer,
          ...child.props
        }
      };
    }
  });
  const shouldEnableDefaultZoom = (0, import_lodash_es.isUndefined)(slotMapToolbarConfig.zoom) && !(0, import_lodash_es.isEmpty)(slotMapToolbarConfig) && slots.toolbar?.hidden !== true;
  if (shouldEnableDefaultZoom) {
    slotMapToolbarConfig.zoom = { enabled: true };
  }
  if (!(0, import_lodash_es.isEmpty)(slotMapToolbarConfig)) {
    slots.toolbar = { ...slots.toolbar, ...slotMapToolbarConfig };
  }
  return slots;
};
const iterateChartInteractionsSlots = (children) => {
  let interactions = {};
  import_react.Children.forEach(children, (childInteractions) => {
    if ((0, import_is_slot.isSlot)(childInteractions, import_ChartInteractions.ChartInteractions.Zoom)) {
      interactions = { ...interactions, zoom: { enabled: true } };
    } else if ((0, import_is_slot.isSlot)(childInteractions, import_ChartInteractions.ChartInteractions.ZoomToFit)) {
      interactions = { ...interactions, zoomToFit: { enabled: true } };
    }
  });
  return interactions;
};
const iterateChartToolbarSlots = (children) => {
  let toolbar = {};
  import_react.Children.forEach(children, (childToolbar) => {
    if ((0, import_is_slot.isSlot)(childToolbar, import_charts.ChartToolbar.DownloadData)) {
      toolbar = {
        ...toolbar,
        downloadCSV: { enabled: true }
      };
    }
  });
  return toolbar;
};
