// @dynatrace/strato-geo/map/utils/tooltip-type-guards.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var tooltip_type_guards_exports = {};
__export(tooltip_type_guards_exports, {
  isBubbleTooltipTemplate: () => isBubbleTooltipTemplate,
  isChoroplethTooltipTemplate: () => isChoroplethTooltipTemplate,
  isConnection: () => isConnection,
  isConnectionTooltipTemplate: () => isConnectionTooltipTemplate,
  isDefaultTooltipHandler: () => import_use_tooltip_template.isDefaultTooltipHandler,
  isDotTooltipTemplate: () => isDotTooltipTemplate,
  isRegion: () => isRegion
});
module.exports = __toCommonJS(tooltip_type_guards_exports);
var import_use_tooltip_template = require("../hooks/use-tooltip-template.js");
var import_BubbleLayerTooltip = require("../slots/BubbleLayerTooltip.js");
var import_ChoroplethLayerTooltip = require("../slots/ChoroplethLayerTooltip.js");
var import_ConnectionLayerTooltip = require("../slots/ConnectionLayerTooltip.js");
var import_DotLayerTooltip = require("../slots/DotLayerTooltip.js");
const isDotTooltipTemplate = (template) => typeof template === typeof import_DotLayerTooltip.DotLayerTooltip;
const isBubbleTooltipTemplate = (template) => typeof template === typeof import_BubbleLayerTooltip.BubbleLayerTooltip;
const isConnectionTooltipTemplate = (template) => typeof template === typeof import_ConnectionLayerTooltip.ConnectionLayerTooltip;
const isChoroplethTooltipTemplate = (template) => typeof template === typeof import_ChoroplethLayerTooltip.ChoroplethLayerTooltip;
const isConnection = (data) => "path" in data.data;
const isRegion = (data) => "name" in data;
