// @dynatrace/strato-components/esm/charts/core/components/interactions-state-helpers/reducers/explore-reducers.js
function reduceExplorePosition(position) {
  return (prev) => ({
    ...prev,
    explore: {
      ...prev.explore,
      visible: true,
      position
    }
  });
}
function reduceExploreEnd(prev) {
  return {
    ...prev,
    explore: {
      ...prev.explore,
      position: prev.explore.pinned ? prev.explore.position : void 0,
      visible: prev.explore.pinned
    }
  };
}
function reduceExploreReset(prev) {
  return {
    ...prev,
    highlightedSeriesId: void 0,
    explore: {
      ...prev.explore,
      visible: false,
      pinned: false,
      position: void 0
    },
    select: void 0
  };
}
function reduceExplorePin(position, highlightedSeriesId) {
  return (prev) => ({
    ...prev,
    highlightedSeriesId,
    explore: {
      ...prev.explore,
      pinned: true,
      position
    }
  });
}
function reduceExploreUnPin(position) {
  return (prev) => ({
    ...prev,
    select: void 0,
    highlightedSeriesId: void 0,
    explore: {
      ...prev.explore,
      pinned: false,
      position
    }
  });
}
function reduceExploreSelect(select) {
  return (prev) => ({
    ...prev,
    select: {
      ...prev.select,
      ...select
    },
    explore: {
      visible: false,
      pinned: false
    },
    highlightedSeriesId: void 0
  });
}
function reduceExploreCancelTooltip(prev) {
  return {
    ...prev,
    explore: {
      visible: false,
      pinned: false
    },
    highlightedSeriesId: void 0
  };
}
function reduceExploreCancelSelect(prev) {
  return {
    ...prev,
    select: void 0
  };
}
function reduceExploreCancelAll(prev) {
  return {
    ...prev,
    explore: {
      visible: false,
      pinned: false
    },
    select: void 0,
    highlightedSeriesId: void 0
  };
}
export {
  reduceExploreCancelAll,
  reduceExploreCancelSelect,
  reduceExploreCancelTooltip,
  reduceExploreEnd,
  reduceExplorePin,
  reduceExplorePosition,
  reduceExploreReset,
  reduceExploreSelect,
  reduceExploreUnPin
};
