// @dynatrace/strato-components/esm/charts/core/components/interactions-state-helpers/reducers/reset-reducers.js
function reduceResetInteractions(isZoomOutEnabled, position) {
  return (prev) => ({
    ...prev,
    highlightedSeriesId: void 0,
    explore: {
      position,
      visible: position !== void 0,
      pinned: false
    },
    select: void 0,
    currentDomain: void 0,
    zoom: {
      zoomInEnabled: true,
      zoomOutEnabled: isZoomOutEnabled
    }
  });
}
function reduceResetZoom(prev) {
  return {
    ...prev,
    currentDomain: void 0,
    highlightedSeriesId: prev.explore.pinned ? void 0 : prev.highlightedSeriesId,
    zoom: {
      zoomOutEnabled: false,
      zoomInEnabled: true
    }
  };
}
function reduceResetToInfiniteZoom(prev) {
  return {
    ...prev,
    currentDomain: void 0,
    highlightedSeriesId: void 0,
    zoom: {
      zoomOutEnabled: true,
      zoomInEnabled: true
    }
  };
}
function reduceResetExploreVisible(prev) {
  return {
    ...prev,
    explore: {
      ...prev.explore,
      visible: false
    }
  };
}
export {
  reduceResetExploreVisible,
  reduceResetInteractions,
  reduceResetToInfiniteZoom,
  reduceResetZoom
};
