// @dynatrace/strato-components/esm/charts/core/components/annotations/hooks/useMarkerEventHandlers.js
import { isUndefined } from "lodash-es";
import { useAnnotationsSelectionState } from "./store/selectors.js";
import {
  useSetStateSelection,
  useSetStateTooltip
} from "./store/state-setters.js";
import { useMarkersInteractions } from "./useMarkersInteractions.js";
import { initialState } from "../providers/AnnotationsDataInitialState.js";
import { calculateTooltipPosition } from "../utils/calculate-tooltip-position.js";
const useMarkerEventHandlers = (annotationsContainerRef) => {
  const updateTooltipState = useSetStateTooltip();
  const updateTooltipSelection = useSetStateSelection();
  const { pinnedGroup } = useAnnotationsSelectionState();
  const onMarkersInteraction = useMarkersInteractions();
  const isTooltipPinned = !isUndefined(pinnedGroup);
  const handleMouseMove = (event) => {
    if (isTooltipPinned || !annotationsContainerRef?.current) {
      return;
    }
    const { clientX, clientY } = event;
    const absolutePosition = calculateTooltipPosition(clientX, clientY);
    const annotationsContainerRect = annotationsContainerRef.current.getBoundingClientRect();
    updateTooltipState({
      position: absolutePosition,
      relativePosition: {
        x: absolutePosition.x - annotationsContainerRect.left,
        y: absolutePosition.y - annotationsContainerRect.top
      }
    });
  };
  const handleMouseEnter = (markerId, groupId, annotationsGroup) => (event) => {
    if (isTooltipPinned || !annotationsContainerRef?.current) {
      return;
    }
    const { clientX, clientY } = event;
    const absolutePosition = calculateTooltipPosition(clientX, clientY);
    const annotationsContainerRect = annotationsContainerRef.current.getBoundingClientRect();
    updateTooltipState({
      position: absolutePosition,
      relativePosition: {
        x: absolutePosition.x - annotationsContainerRect.left,
        y: absolutePosition.y - annotationsContainerRect.top
      },
      data: annotationsGroup,
      visibility: true
    });
    const selectionState = {
      hoveredItem: markerId,
      hoveredGroup: groupId,
      data: annotationsGroup
    };
    updateTooltipSelection(selectionState);
    onMarkersInteraction?.(selectionState);
  };
  const handleClick = (groupId, annotationsGroup) => (event) => {
    event.stopPropagation();
    const { clientX, clientY } = event;
    const newSelection = {
      hoveredGroup: groupId,
      data: annotationsGroup,
      pinnedGroup: isTooltipPinned ? void 0 : groupId
      // if tooltip was already pinned, unpin
    };
    updateTooltipState({
      position: calculateTooltipPosition(clientX, clientY),
      data: annotationsGroup,
      visibility: true,
      pinned: !isTooltipPinned
      // if tooltip was already pinned, show updated info but unpin
    });
    updateTooltipSelection(newSelection);
    onMarkersInteraction?.(newSelection);
  };
  const handleMouseLeave = () => {
    if (isTooltipPinned) {
      return;
    }
    updateTooltipState(initialState.tooltip);
    updateTooltipSelection(initialState.selection);
    onMarkersInteraction?.(initialState.selection);
  };
  return {
    handleMouseEnter,
    handleMouseMove,
    handleMouseLeave,
    handleClick
  };
};
export {
  useMarkerEventHandlers
};
