// @dynatrace/strato-components/charts/categorical-bar/layout/CategoricalGridZeroLineLayout.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CategoricalGridZeroLineLayout_exports = {};
__export(CategoricalGridZeroLineLayout_exports, {
  CategoricalGridZeroLineLayout: () => CategoricalGridZeroLineLayout
});
module.exports = __toCommonJS(CategoricalGridZeroLineLayout_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_ChartGrid_css = require("../../core/components/chart-grid/ChartGrid.sty.js");
var import_useGridPlotSize = require("../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_GridLine = require("../../core/components/grid/GridLine.js");
var import_useCategoricalBarChartBaseConfig = require("../hooks/useCategoricalBarChartBaseConfig.js");
var import_useCategoricalBarChartScales = require("../hooks/useCategoricalBarChartScales.js");
var import_build_zero_line = require("../utils/build-zero-line.js");
var import_layout = require("../utils/layout.js");
const CategoricalGridZeroLineLayout = () => {
  const { layout } = (0, import_useCategoricalBarChartBaseConfig.useCategoricalBarChartBaseConfig)();
  const { valueScale: scale } = (0, import_useCategoricalBarChartScales.useCategoricalBarChartScales)();
  const { width, height } = (0, import_useGridPlotSize.useGridPlotSize)();
  const zeroLineValue = scale(0) ?? 0;
  const referenceSize = (0, import_layout.isVertical)(layout) ? height : width;
  if (zeroLineValue === referenceSize || zeroLineValue === 0) {
    return null;
  }
  const {
    show: showLine,
    from,
    to
  } = (0, import_build_zero_line.buildZeroLine)(layout, zeroLineValue, width, height);
  return showLine ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "svg",
    {
      className: import_ChartGrid_css.plotLayout,
      role: "graphics-symbol",
      "aria-label": "categorical grid zero line",
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_GridLine.GridLine, { from, to, highlight: true })
    }
  ) : null;
};
CategoricalGridZeroLineLayout["displayName"] = "CategoricalGridZeroLineLayout";
