// @dynatrace/strato-components/charts/honeycomb/components/tooltip/HoneycombOverlayTooltip.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var HoneycombOverlayTooltip_exports = {};
__export(HoneycombOverlayTooltip_exports, {
  HoneycombOverlayTooltip: () => HoneycombOverlayTooltip
});
module.exports = __toCommonJS(HoneycombOverlayTooltip_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_utils = require("../../../core/components/overlay-tooltip/_to-be-removed/utils.js");
var import_OverlayTooltip = require("../../../core/components/overlay-tooltip/OverlayTooltip.js");
var import_utils2 = require("../../../core/components/overlay-tooltip/utils.js");
var import_use_honeycomb_custom_tooltip = require("../../hooks/use-honeycomb-custom-tooltip.js");
var import_use_honeycomb_truncation_mode = require("../../hooks/use-honeycomb-truncation-mode.js");
const HoneycombOverlayTooltip = (props) => {
  const data = (0, import_utils.useTooltipData)().data;
  const truncationMode = (0, import_use_honeycomb_truncation_mode.useHoneycombTruncationMode)();
  const tooltip = (0, import_use_honeycomb_custom_tooltip.useHoneycombCustomTooltip)();
  if (tooltip?.hidden || (0, import_utils2.isTooltipDataEmpty)(data)) {
    return null;
  }
  const series = data.series[0];
  if (typeof tooltip?.template === "function") {
    const metadata = data.metadata;
    if (metadata) {
      const payload = {
        name: metadata.tooltipMetadata.name,
        value: metadata.tileData.value,
        formattedValue: String(series.value),
        color: metadata.tooltipMetadata.color.Default,
        data: metadata.tileData
      };
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayTooltip.OverlayTooltip, { legacyTemplate: tooltip.template(payload) });
    }
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayTooltip.OverlayTooltip, { truncation: truncationMode, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_OverlayTooltip.OverlayTooltip.SeriesItem,
    {
      name: series.name,
      value: series.value,
      color: series.color,
      actions: props.actionsForItem?.(series)
    }
  ) });
};
HoneycombOverlayTooltip["displayName"] = "HoneycombOverlayTooltip";
