// @dynatrace/strato-components/charts/core/components/grid/Grid.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Grid_exports = {};
__export(Grid_exports, {
  Grid: () => Grid
});
module.exports = __toCommonJS(Grid_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_group = require("@visx/group");
var import_GridColumns = require("../../../core/components/grid/GridColumns.js");
var import_GridRows = require("../../../core/components/grid/GridRows.js");
var import_ticks = require("../../../core/utils/ticks.js");
var import_detect_scale = require("../../utils/detect-scale.js");
const DEFAULTS = {
  HIGHLIGHT_ZERO_LINE: {
    horizontal: true,
    vertical: true
  }
};
function Grid({
  width,
  height,
  xScale,
  yScale,
  highlightZeroLine = DEFAULTS.HIGHLIGHT_ZERO_LINE,
  xTicks,
  yTicks,
  xCategoryOffset,
  yCategoryOffset
}) {
  const numTicksRows = (0, import_ticks.getVerticalNumTicks)(height);
  const autoNumTicks = (0, import_ticks.getHorizontalNumTicks)(width);
  const xNumTicks = xCategoryOffset !== void 0 && (0, import_detect_scale.isBandScale)(xScale) ? xScale.domain().length : autoNumTicks;
  const yNumTicks = yCategoryOffset !== void 0 && (0, import_detect_scale.isBandScale)(yScale) ? yScale.domain().length : numTicksRows;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_group.Group, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_GridColumns.GridColumns,
      {
        width,
        height,
        scale: xScale,
        numTicks: xNumTicks,
        highlightZeroLine: highlightZeroLine.vertical,
        ticks: xTicks,
        offset: xCategoryOffset
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_GridRows.GridRows,
      {
        width,
        height,
        scale: yScale,
        numTicks: yNumTicks,
        highlightZeroLine: highlightZeroLine.horizontal,
        ticks: yTicks,
        offset: yCategoryOffset
      }
    )
  ] });
}
