// @dynatrace/strato-components/charts/core/components/grid/GridLine.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var GridLine_exports = {};
__export(GridLine_exports, {
  DEFAULT_STROKE_COLOR: () => DEFAULT_STROKE_COLOR,
  DEFAULT_STROKE_WIDTH: () => DEFAULT_STROKE_WIDTH,
  GridLine: () => GridLine,
  HIGHLIGHTED_STROKE_COLOR: () => HIGHLIGHTED_STROKE_COLOR,
  HIGHLIGHTED_STROKE_WIDTH: () => HIGHLIGHTED_STROKE_WIDTH
});
module.exports = __toCommonJS(GridLine_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_shape = require("@visx/shape");
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
const DEFAULT_STROKE_WIDTH = "1px";
const DEFAULT_STROKE_COLOR = import_colors.default.Background.Field.Neutral.DefaultActive;
const HIGHLIGHTED_STROKE_WIDTH = "2px";
const HIGHLIGHTED_STROKE_COLOR = import_colors.default.Text.Neutral.Subdued;
const GridLine = ({ from, to, highlight = false }) => {
  let stroke = DEFAULT_STROKE_COLOR;
  let strokeWidth = DEFAULT_STROKE_WIDTH;
  if (highlight) {
    stroke = HIGHLIGHTED_STROKE_COLOR;
    strokeWidth = HIGHLIGHTED_STROKE_WIDTH;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_shape.Line,
    {
      from,
      to,
      stroke,
      strokeWidth,
      style: {
        shapeRendering: "geometricPrecision"
      }
    }
  );
};
