// @dynatrace-sdk/client-platform-management-service/esm/index.js
/**
 * @license
 * Copyright 2023-2025 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// packages/client/platform-management-service/src/lib/apis/effective-permissions-api.ts
import {
  httpClient as defaultHttpClient,
  isHttpClientAbortError,
  isHttpClientResponseError
} from "@dynatrace-sdk/http-client";
import {
  apiGatewayErrorsHandler,
  getErrorMessage,
  transformRequest,
  transformResponse
} from "@dynatrace-sdk/shared-client-utils";
import {
  ApiClientError,
  InvalidResponseError,
  isInvalidResponseError
} from "@dynatrace-sdk/shared-errors";

// packages/client/platform-management-service/src/lib/error-envelopes/error-envelope-error.ts
import { ClientRequestError } from "@dynatrace-sdk/shared-errors";
var ErrorEnvelopeError = class extends ClientRequestError {
  isErrorEnvelopeError = true;
};
function isErrorEnvelopeError(e) {
  return e?.isApiClientError === true && e?.isClientRequestError === true && e?.isErrorEnvelopeError === true && e instanceof Error;
}

// packages/client/platform-management-service/src/lib/error-envelopes/responses/default-error-response.ts
var DefaultErrorResponse = class extends ErrorEnvelopeError {
  isDefaultErrorResponse = true;
  name = "DefaultErrorResponse";
};
function isDefaultErrorResponse(e) {
  return e?.isApiClientError === true && e?.isClientRequestError === true && e?.isErrorEnvelopeError === true && e?.isDefaultErrorResponse === true && e instanceof Error;
}

// packages/client/platform-management-service/src/lib/apis/effective-permissions-api.ts
var EffectivePermissionsClient = class {
  httpClient;
  shouldTransformDates = true;
  /**
   * @param {HttpClient} httpClientImplementation - You can provide custom http client as a parameter to constructor of a EffectivePermissionsClient. Custom http client should implement {HttpClient} interface.
   * @param {ApiClientOptions} [options] - Optional configuration for SDK Client.
   * @example
   * const effectivePermissionsClientNodeJs = new EffectivePermissionsClient(yourCustomImplementation);
   */
  constructor(httpClientImplementation, options) {
    this.httpClient = httpClientImplementation;
    this.shouldTransformDates = options?.transformDates ?? true;
  }
  /**
   * Get the effective permissions of the calling user in the environment
   *
   * **One of the following scopes is required:**
   * + app-engine:apps:run
   * + app-engine:functions:run
   *
   * <br/><br/>This endpoint resolves whether the caller (based on the provided OAuth token when calling this endpoint) has the requested permissions (see [IAM policy reference](https://dt-url.net/ut02qdv) for a list of permissions and their conditions). It resolves the permission via IAM and uses the optional context to evaluate conditional results. The provided context consists of key-value pairs where the `key` must correspond to the name of a supported condition of the requested permission and the `value` is the value that it this condition is evaluated against given a user's policies.
   *
   * For each requested permission the response indicates whether the user has the requested permission (`"granted": "true"`), does not have it (`"granted": "false"`) or only has it conditionally (`"granted": "condition"`), meaning that the provided values for conditions in the context are not sufficient to decide whether the caller has the permission.
   *
   * For environment API v2 scopes (i.e. scopes starting with environment-api:) roles on the environment and management zone level will also be considered.
   *
   * > ℹ️ **Settings permission** requests for scopes `settings:schemas:read`, `settings:objects:read`, and `settings:objects:write` will always result in a result with `"granted": "condition"`. You can use the dedicated [effective permissions endpoint for settings](/platform/swagger-ui/index.html?urls.primaryName=Classic+Environment+API+v2+%28for+SDK+generation%29) instead.
   *
   * > ℹ️ Apps must list all permissions they check in the list of `scopes` in their app configuration.
   *
   * You can learn more about querying user permissions on [Dynatrace Developer](https://dt-url.net/query-user-permissions).
   *
   * @returns {Promise<EffectivePermissions>} Success
   *
   * @example <caption>Code example</caption>
   * import { effectivePermissionsClient } from "@dynatrace-sdk/client-platform-management-service";
   *
   * const data =
   *   await effectivePermissionsClient.resolveEffectivePermissions(
   *     {
   *       body: {
   *         permissions: [
   *           { permission: "state:app-states:write" },
   *         ],
   *       },
   *     },
   *   );
   *
   * @throws {DefaultErrorResponse} Unexpected error
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async resolveEffectivePermissions(config) {
    if (!config) {
      throw new ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const encodedBody = this.shouldTransformDates ? transformRequest(config.body, []) : config.body;
    try {
      const response = await this.httpClient.send({
        url: `/platform/management/v1/effective-permissions:resolve`,
        method: "POST",
        requestBodyType: "json",
        body: encodedBody,
        headers: {
          "Content-Type": "application/json",
          Accept: "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
      } catch (err) {
        throw new InvalidResponseError(
          `EffectivePermissionsClient.resolveEffectivePermissions:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if (isHttpClientAbortError(e) || isInvalidResponseError(e)) {
        throw e;
      }
      if (!isHttpClientResponseError(e)) {
        throw new ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      const responseValue = await response.body("json");
      try {
        const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
        throw new DefaultErrorResponse(`err`, response, errorBody, getErrorMessage(errorBody, "Unexpected error"), e);
      } catch (err) {
        if (err instanceof DefaultErrorResponse) {
          throw err;
        }
        throw new InvalidResponseError(
          `EffectivePermissionsClient.resolveEffectivePermissions:err`,
          err,
          responseValue,
          "ErrorEnvelope",
          void 0,
          response
        );
      }
    }
  }
};
var effectivePermissionsClient = /* @__PURE__ */ new EffectivePermissionsClient(defaultHttpClient);

// packages/client/platform-management-service/src/lib/apis/environment-information-api.ts
import {
  httpClient as defaultHttpClient2,
  isHttpClientAbortError as isHttpClientAbortError2,
  isHttpClientResponseError as isHttpClientResponseError2
} from "@dynatrace-sdk/http-client";
import {
  apiGatewayErrorsHandler as apiGatewayErrorsHandler2,
  getErrorMessage as getErrorMessage2,
  transformResponse as transformResponse2
} from "@dynatrace-sdk/shared-client-utils";
import {
  ApiClientError as ApiClientError2,
  InvalidResponseError as InvalidResponseError2,
  isInvalidResponseError as isInvalidResponseError2
} from "@dynatrace-sdk/shared-errors";

// packages/client/platform-management-service/src/lib/error-envelopes/responses/bad-request.ts
var BadRequest = class extends ErrorEnvelopeError {
  isBadRequest = true;
  name = "BadRequest";
};
function isBadRequest(e) {
  return e?.isApiClientError === true && e?.isClientRequestError === true && e?.isErrorEnvelopeError === true && e?.isBadRequest === true && e instanceof Error;
}

// packages/client/platform-management-service/src/lib/error-envelopes/responses/not-found.ts
var NotFound = class extends ErrorEnvelopeError {
  isNotFound = true;
  name = "NotFound";
};
function isNotFound(e) {
  return e?.isApiClientError === true && e?.isClientRequestError === true && e?.isErrorEnvelopeError === true && e?.isNotFound === true && e instanceof Error;
}

// packages/client/platform-management-service/src/lib/error-envelopes/responses/service-unavailable.ts
var ServiceUnavailable = class extends ErrorEnvelopeError {
  isServiceUnavailable = true;
  name = "ServiceUnavailable";
};
function isServiceUnavailable(e) {
  return e?.isApiClientError === true && e?.isClientRequestError === true && e?.isErrorEnvelopeError === true && e?.isServiceUnavailable === true && e instanceof Error;
}

// packages/client/platform-management-service/src/lib/error-envelopes/responses/unauthorized.ts
var Unauthorized = class extends ErrorEnvelopeError {
  isUnauthorized = true;
  name = "Unauthorized";
};
function isUnauthorized(e) {
  return e?.isApiClientError === true && e?.isClientRequestError === true && e?.isErrorEnvelopeError === true && e?.isUnauthorized === true && e instanceof Error;
}

// packages/client/platform-management-service/src/lib/apis/environment-information-api.ts
var EnvironmentInformationClient = class {
  httpClient;
  shouldTransformDates = true;
  /**
   * @param {HttpClient} httpClientImplementation - You can provide custom http client as a parameter to constructor of a EnvironmentInformationClient. Custom http client should implement {HttpClient} interface.
   * @param {ApiClientOptions} [options] - Optional configuration for SDK Client.
   * @example
   * const environmentInformationClientNodeJs = new EnvironmentInformationClient(yourCustomImplementation);
   */
  constructor(httpClientImplementation, options) {
    this.httpClient = httpClientImplementation;
    this.shouldTransformDates = options?.transformDates ?? true;
  }
  /**
   * Get basic environment information
   *
   * **One of the following scopes is required:**
   * + app-engine:apps:run
   * + app-engine:functions:run
   *
   * <br/><br/>Get basic information about the current environment.
   *
   * @returns {Promise<EnvironmentInfo>} Info about the requested environment.
   *
   * @example <caption>Code example</caption>
   * import { environmentInformationClient } from "@dynatrace-sdk/client-platform-management-service";
   *
   * const data =
   *   await environmentInformationClient.getEnvironmentInformation();
   *
   * @throws {BadRequest} Bad request
   * @throws {Unauthorized} Authentication failed
   * @throws {NotFound} Resource not found
   * @throws {ServiceUnavailable} There is a temporary problem in the backend.
   * @throws {DefaultErrorResponse} Unexpected error
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async getEnvironmentInformation(config = {}) {
    try {
      const response = await this.httpClient.send({
        url: `/platform/management/v1/environment`,
        method: "GET",
        headers: {
          Accept: "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? transformResponse2(responseValue, [
          ["createTime", { datetime: true }],
          ["blockTime", { datetime: true }]
        ]) : responseValue;
      } catch (err) {
        throw new InvalidResponseError2(
          `EnvironmentInformationClient.getEnvironmentInformation:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if (isHttpClientAbortError2(e) || isInvalidResponseError2(e)) {
        throw e;
      }
      if (!isHttpClientResponseError2(e)) {
        throw new ApiClientError2("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler2(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse2(responseValue, []) : responseValue;
            throw new BadRequest(`400`, response, errorBody, getErrorMessage2(errorBody, "Bad request"), e);
          } catch (err) {
            if (err instanceof BadRequest) {
              throw err;
            }
            throw new InvalidResponseError2(
              `EnvironmentInformationClient.getEnvironmentInformation:400`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 401: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse2(responseValue, []) : responseValue;
            throw new Unauthorized(`401`, response, errorBody, getErrorMessage2(errorBody, "Authentication failed"), e);
          } catch (err) {
            if (err instanceof Unauthorized) {
              throw err;
            }
            throw new InvalidResponseError2(
              `EnvironmentInformationClient.getEnvironmentInformation:401`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse2(responseValue, []) : responseValue;
            throw new NotFound(`404`, response, errorBody, getErrorMessage2(errorBody, "Resource not found"), e);
          } catch (err) {
            if (err instanceof NotFound) {
              throw err;
            }
            throw new InvalidResponseError2(
              `EnvironmentInformationClient.getEnvironmentInformation:404`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 503: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse2(responseValue, []) : responseValue;
            throw new ServiceUnavailable(
              `503`,
              response,
              errorBody,
              getErrorMessage2(errorBody, "There is a temporary problem in the backend."),
              e
            );
          } catch (err) {
            if (err instanceof ServiceUnavailable) {
              throw err;
            }
            throw new InvalidResponseError2(
              `EnvironmentInformationClient.getEnvironmentInformation:503`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        default: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse2(responseValue, []) : responseValue;
            throw new DefaultErrorResponse(
              `err`,
              response,
              errorBody,
              getErrorMessage2(errorBody, "Unexpected error"),
              e
            );
          } catch (err) {
            if (err instanceof DefaultErrorResponse) {
              throw err;
            }
            throw new InvalidResponseError2(
              `EnvironmentInformationClient.getEnvironmentInformation:err`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
      }
    }
  }
};
var environmentInformationClient = /* @__PURE__ */ new EnvironmentInformationClient(defaultHttpClient2);

// packages/client/platform-management-service/src/lib/apis/environment-settings-api.ts
import {
  httpClient as defaultHttpClient3,
  isHttpClientAbortError as isHttpClientAbortError3,
  isHttpClientResponseError as isHttpClientResponseError3
} from "@dynatrace-sdk/http-client";
import {
  apiGatewayErrorsHandler as apiGatewayErrorsHandler3,
  getErrorMessage as getErrorMessage3,
  transformResponse as transformResponse3
} from "@dynatrace-sdk/shared-client-utils";
import {
  ApiClientError as ApiClientError3,
  InvalidResponseError as InvalidResponseError3,
  isInvalidResponseError as isInvalidResponseError3
} from "@dynatrace-sdk/shared-errors";
var EnvironmentSettingsClient = class {
  httpClient;
  shouldTransformDates = true;
  /**
   * @param {HttpClient} httpClientImplementation - You can provide custom http client as a parameter to constructor of a EnvironmentSettingsClient. Custom http client should implement {HttpClient} interface.
   * @param {ApiClientOptions} [options] - Optional configuration for SDK Client.
   * @example
   * const environmentSettingsClientNodeJs = new EnvironmentSettingsClient(yourCustomImplementation);
   */
  constructor(httpClientImplementation, options) {
    this.httpClient = httpClientImplementation;
    this.shouldTransformDates = options?.transformDates ?? true;
  }
  /**
   * Get settings for the environment
   *
   * **One of the following scopes is required:**
   * + app-engine:apps:run
   * + app-engine:functions:run
   *
   * <br/><br/>Gets settings for environment.
   *
   * @returns {Promise<SettingsResponse>} The settings for the requested environment
   *
   * @example <caption>Code example</caption>
   * import { environmentSettingsClient } from "@dynatrace-sdk/client-platform-management-service";
   *
   * const data =
   *   await environmentSettingsClient.getEnvironmentSettings();
   *
   * @throws {BadRequest} Bad request
   * @throws {Unauthorized} Authentication failed
   * @throws {NotFound} Resource not found
   * @throws {ServiceUnavailable} There is a temporary problem in the backend.
   * @throws {DefaultErrorResponse} Unexpected error
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async getEnvironmentSettings(config = {}) {
    try {
      const response = await this.httpClient.send({
        url: `/platform/management/v1/environment/settings`,
        method: "GET",
        headers: {
          Accept: "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? transformResponse3(responseValue, [["activeTrialSubscriptionEndDate", { datetime: true }]]) : responseValue;
      } catch (err) {
        throw new InvalidResponseError3(
          `EnvironmentSettingsClient.getEnvironmentSettings:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if (isHttpClientAbortError3(e) || isInvalidResponseError3(e)) {
        throw e;
      }
      if (!isHttpClientResponseError3(e)) {
        throw new ApiClientError3("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler3(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse3(responseValue, []) : responseValue;
            throw new BadRequest(`400`, response, errorBody, getErrorMessage3(errorBody, "Bad request"), e);
          } catch (err) {
            if (err instanceof BadRequest) {
              throw err;
            }
            throw new InvalidResponseError3(
              `EnvironmentSettingsClient.getEnvironmentSettings:400`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 401: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse3(responseValue, []) : responseValue;
            throw new Unauthorized(`401`, response, errorBody, getErrorMessage3(errorBody, "Authentication failed"), e);
          } catch (err) {
            if (err instanceof Unauthorized) {
              throw err;
            }
            throw new InvalidResponseError3(
              `EnvironmentSettingsClient.getEnvironmentSettings:401`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse3(responseValue, []) : responseValue;
            throw new NotFound(`404`, response, errorBody, getErrorMessage3(errorBody, "Resource not found"), e);
          } catch (err) {
            if (err instanceof NotFound) {
              throw err;
            }
            throw new InvalidResponseError3(
              `EnvironmentSettingsClient.getEnvironmentSettings:404`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 503: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse3(responseValue, []) : responseValue;
            throw new ServiceUnavailable(
              `503`,
              response,
              errorBody,
              getErrorMessage3(errorBody, "There is a temporary problem in the backend."),
              e
            );
          } catch (err) {
            if (err instanceof ServiceUnavailable) {
              throw err;
            }
            throw new InvalidResponseError3(
              `EnvironmentSettingsClient.getEnvironmentSettings:503`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        default: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse3(responseValue, []) : responseValue;
            throw new DefaultErrorResponse(
              `err`,
              response,
              errorBody,
              getErrorMessage3(errorBody, "Unexpected error"),
              e
            );
          } catch (err) {
            if (err instanceof DefaultErrorResponse) {
              throw err;
            }
            throw new InvalidResponseError3(
              `EnvironmentSettingsClient.getEnvironmentSettings:err`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
      }
    }
  }
};
var environmentSettingsClient = /* @__PURE__ */ new EnvironmentSettingsClient(defaultHttpClient3);

// packages/client/platform-management-service/src/lib/apis/license-information-api.ts
import {
  httpClient as defaultHttpClient4,
  isHttpClientAbortError as isHttpClientAbortError4,
  isHttpClientResponseError as isHttpClientResponseError4
} from "@dynatrace-sdk/http-client";
import {
  apiGatewayErrorsHandler as apiGatewayErrorsHandler4,
  getErrorMessage as getErrorMessage4,
  toQueryString,
  transformResponse as transformResponse4
} from "@dynatrace-sdk/shared-client-utils";
import {
  ApiClientError as ApiClientError4,
  InvalidResponseError as InvalidResponseError4,
  isInvalidResponseError as isInvalidResponseError4
} from "@dynatrace-sdk/shared-errors";
var LicenseInformationClient = class {
  httpClient;
  shouldTransformDates = true;
  /**
   * @param {HttpClient} httpClientImplementation - You can provide custom http client as a parameter to constructor of a LicenseInformationClient. Custom http client should implement {HttpClient} interface.
   * @param {ApiClientOptions} [options] - Optional configuration for SDK Client.
   * @example
   * const licenseInformationClientNodeJs = new LicenseInformationClient(yourCustomImplementation);
   */
  constructor(httpClientImplementation, options) {
    this.httpClient = httpClientImplementation;
    this.shouldTransformDates = options?.transformDates ?? true;
  }
  /**
   * Get basic license information
   *
   * **One of the following scopes is required:**
   * + app-engine:apps:run
   * + app-engine:functions:run
   *
   * <br/><br/>Get basic license information about the current environment.
   *
   * @returns {Promise<License>} The license info of the requested environment.
   *
   * @example <caption>Code example</caption>
   * import { licenseInformationClient } from "@dynatrace-sdk/client-platform-management-service";
   *
   * const data = await licenseInformationClient.getLicense();
   *
   * @throws {BadRequest} Bad request
   * @throws {Unauthorized} Authentication failed
   * @throws {NotFound} Resource not found
   * @throws {ServiceUnavailable} There is a temporary problem in the backend.
   * @throws {DefaultErrorResponse} Unexpected error
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async getLicense(config = {}) {
    try {
      const response = await this.httpClient.send({
        url: `/platform/management/v1/environment/license`,
        method: "GET",
        headers: {
          Accept: "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? transformResponse4(responseValue, []) : responseValue;
      } catch (err) {
        throw new InvalidResponseError4(
          `LicenseInformationClient.getLicense:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if (isHttpClientAbortError4(e) || isInvalidResponseError4(e)) {
        throw e;
      }
      if (!isHttpClientResponseError4(e)) {
        throw new ApiClientError4("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler4(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse4(responseValue, []) : responseValue;
            throw new BadRequest(`400`, response, errorBody, getErrorMessage4(errorBody, "Bad request"), e);
          } catch (err) {
            if (err instanceof BadRequest) {
              throw err;
            }
            throw new InvalidResponseError4(
              `LicenseInformationClient.getLicense:400`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 401: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse4(responseValue, []) : responseValue;
            throw new Unauthorized(`401`, response, errorBody, getErrorMessage4(errorBody, "Authentication failed"), e);
          } catch (err) {
            if (err instanceof Unauthorized) {
              throw err;
            }
            throw new InvalidResponseError4(
              `LicenseInformationClient.getLicense:401`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse4(responseValue, []) : responseValue;
            throw new NotFound(`404`, response, errorBody, getErrorMessage4(errorBody, "Resource not found"), e);
          } catch (err) {
            if (err instanceof NotFound) {
              throw err;
            }
            throw new InvalidResponseError4(
              `LicenseInformationClient.getLicense:404`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 503: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse4(responseValue, []) : responseValue;
            throw new ServiceUnavailable(
              `503`,
              response,
              errorBody,
              getErrorMessage4(errorBody, "There is a temporary problem in the backend."),
              e
            );
          } catch (err) {
            if (err instanceof ServiceUnavailable) {
              throw err;
            }
            throw new InvalidResponseError4(
              `LicenseInformationClient.getLicense:503`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        default: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse4(responseValue, []) : responseValue;
            throw new DefaultErrorResponse(
              `err`,
              response,
              errorBody,
              getErrorMessage4(errorBody, "Unexpected error"),
              e
            );
          } catch (err) {
            if (err instanceof DefaultErrorResponse) {
              throw err;
            }
            throw new InvalidResponseError4(
              `LicenseInformationClient.getLicense:err`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
      }
    }
  }
  /**
   * Get basic license settings information
   *
   * **One of the following scopes is required:**
   * + app-engine:apps:run
   * + app-engine:functions:run
   *
   * <br/><br/>Get basic license settings information about the current environment
   *
   * @returns {Promise<LicenseSettingsResponse>} The license settings info of the requested environment.
   *
   * @example <caption>Code example</caption>
   * import { licenseInformationClient } from "@dynatrace-sdk/client-platform-management-service";
   *
   * const data =
   *   await licenseInformationClient.getLicenseSettings();
   *
   * @throws {BadRequest} Bad request
   * @throws {Unauthorized} Authentication failed
   * @throws {NotFound} Resource not found
   * @throws {ServiceUnavailable} There is a temporary problem in the backend.
   * @throws {DefaultErrorResponse} Unexpected error
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async getLicenseSettings(config = {}) {
    if (!config) {
      throw new ApiClientError4("API client error", "API client call is missing mandatory config parameter");
    }
    const query = toQueryString({ keys: config.keys });
    try {
      const response = await this.httpClient.send({
        url: `/platform/management/v1/environment/license/settings${query}`,
        method: "GET",
        headers: {
          Accept: "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? transformResponse4(responseValue, []) : responseValue;
      } catch (err) {
        throw new InvalidResponseError4(
          `LicenseInformationClient.getLicenseSettings:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if (isHttpClientAbortError4(e) || isInvalidResponseError4(e)) {
        throw e;
      }
      if (!isHttpClientResponseError4(e)) {
        throw new ApiClientError4("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler4(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse4(responseValue, []) : responseValue;
            throw new BadRequest(`400`, response, errorBody, getErrorMessage4(errorBody, "Bad request"), e);
          } catch (err) {
            if (err instanceof BadRequest) {
              throw err;
            }
            throw new InvalidResponseError4(
              `LicenseInformationClient.getLicenseSettings:400`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 401: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse4(responseValue, []) : responseValue;
            throw new Unauthorized(`401`, response, errorBody, getErrorMessage4(errorBody, "Authentication failed"), e);
          } catch (err) {
            if (err instanceof Unauthorized) {
              throw err;
            }
            throw new InvalidResponseError4(
              `LicenseInformationClient.getLicenseSettings:401`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse4(responseValue, []) : responseValue;
            throw new NotFound(`404`, response, errorBody, getErrorMessage4(errorBody, "Resource not found"), e);
          } catch (err) {
            if (err instanceof NotFound) {
              throw err;
            }
            throw new InvalidResponseError4(
              `LicenseInformationClient.getLicenseSettings:404`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 503: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse4(responseValue, []) : responseValue;
            throw new ServiceUnavailable(
              `503`,
              response,
              errorBody,
              getErrorMessage4(errorBody, "There is a temporary problem in the backend."),
              e
            );
          } catch (err) {
            if (err instanceof ServiceUnavailable) {
              throw err;
            }
            throw new InvalidResponseError4(
              `LicenseInformationClient.getLicenseSettings:503`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        default: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse4(responseValue, []) : responseValue;
            throw new DefaultErrorResponse(
              `err`,
              response,
              errorBody,
              getErrorMessage4(errorBody, "Unexpected error"),
              e
            );
          } catch (err) {
            if (err instanceof DefaultErrorResponse) {
              throw err;
            }
            throw new InvalidResponseError4(
              `LicenseInformationClient.getLicenseSettings:err`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
      }
    }
  }
};
var licenseInformationClient = /* @__PURE__ */ new LicenseInformationClient(defaultHttpClient4);

// packages/client/platform-management-service/src/lib/error-envelopes/index.ts
import {
  ApiClientError as ApiClientError5,
  ApiGatewayError,
  ClientRequestError as ClientRequestError6,
  InvalidResponseError as InvalidResponseError5,
  isApiClientError,
  isApiGatewayError,
  isClientRequestError,
  isInvalidResponseError as isInvalidResponseError5
} from "@dynatrace-sdk/shared-errors";

// packages/client/platform-management-service/src/lib/models/effective-permission-granted.ts
var EffectivePermissionGranted = /* @__PURE__ */ ((EffectivePermissionGranted2) => {
  EffectivePermissionGranted2["True"] = "true";
  EffectivePermissionGranted2["False"] = "false";
  EffectivePermissionGranted2["Condition"] = "condition";
  return EffectivePermissionGranted2;
})(EffectivePermissionGranted || {});

// packages/client/platform-management-service/src/lib/models/environment-info-type.ts
var EnvironmentInfoType = /* @__PURE__ */ ((EnvironmentInfoType2) => {
  EnvironmentInfoType2["Internal"] = "INTERNAL";
  EnvironmentInfoType2["Customer"] = "CUSTOMER";
  EnvironmentInfoType2["SelfMonitoring"] = "SELF_MONITORING";
  return EnvironmentInfoType2;
})(EnvironmentInfoType || {});

// packages/client/platform-management-service/src/lib/models/environment-state.ts
var EnvironmentState = /* @__PURE__ */ ((EnvironmentState2) => {
  EnvironmentState2["BeingCreated"] = "BEING_CREATED";
  EnvironmentState2["CreatedNotInitialized"] = "CREATED_NOT_INITIALIZED";
  EnvironmentState2["Active"] = "ACTIVE";
  EnvironmentState2["Deactivated"] = "DEACTIVATED";
  EnvironmentState2["BeingDeleted"] = "BEING_DELETED";
  EnvironmentState2["Blocked"] = "BLOCKED";
  EnvironmentState2["DeletionFailed"] = "DELETION_FAILED";
  EnvironmentState2["PreAllocated"] = "PRE_ALLOCATED";
  return EnvironmentState2;
})(EnvironmentState || {});
export {
  ApiClientError5 as ApiClientError,
  ApiGatewayError,
  BadRequest,
  ClientRequestError6 as ClientRequestError,
  DefaultErrorResponse,
  EffectivePermissionGranted,
  EffectivePermissionsClient,
  EnvironmentInfoType,
  EnvironmentInformationClient,
  EnvironmentSettingsClient,
  EnvironmentState,
  ErrorEnvelopeError,
  InvalidResponseError5 as InvalidResponseError,
  LicenseInformationClient,
  NotFound,
  ServiceUnavailable,
  Unauthorized,
  effectivePermissionsClient,
  environmentInformationClient,
  environmentSettingsClient,
  isApiClientError,
  isApiGatewayError,
  isBadRequest,
  isClientRequestError,
  isDefaultErrorResponse,
  isErrorEnvelopeError,
  isInvalidResponseError5 as isInvalidResponseError,
  isNotFound,
  isServiceUnavailable,
  isUnauthorized,
  licenseInformationClient
};
