// @dynatrace/strato-components/filters/timeframe-selector/components/TimeframeSelectorTooltip.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TimeframeSelectorTooltip_exports = {};
__export(TimeframeSelectorTooltip_exports, {
  TimeframeSelectorTooltip: () => TimeframeSelectorTooltip
});
module.exports = __toCommonJS(TimeframeSelectorTooltip_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_Flex = require("../../../layouts/flex/Flex.js");
var import_Tooltip = require("../../../overlays/tooltip/Tooltip.js");
var import_Text = require("../../../typography/text/Text.js");
var import_useTooltipMessages = require("../hooks/useTooltipMessages.js");
const TimeframeSelectorTooltip = (props) => {
  const {
    value,
    disabled,
    displayValue,
    children
  } = props;
  const [tooltipTimeValues, setTooltipTimeValues] = (0, import_react.useState)({
    tooltipFormattedTimeValueStart: "",
    tooltipFormattedTimeValueEnd: ""
  });
  const {
    tooltipFormattedTimeValueStart,
    tooltipFormattedTimeValueEnd,
    timeframeDuration,
    isAbsoluteTimeframe,
    isInvalid
  } = (0, import_useTooltipMessages.useTooltipMessages)(value);
  const onOpenTooltip = (0, import_react.useCallback)(
    (open) => {
      if (open) {
        setTooltipTimeValues({
          tooltipFormattedTimeValueStart,
          tooltipFormattedTimeValueEnd
        });
      }
    },
    [tooltipFormattedTimeValueEnd, tooltipFormattedTimeValueStart]
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Tooltip.Tooltip,
    {
      text: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { flexDirection: "column", gap: 4, children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, {
          gap: 0,
          flexDirection: "column",
          // displayValue is a string when a preset item name is used.
          children: [
            (typeof displayValue === "string" || !isAbsoluteTimeframe) && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { textStyle: "small-emphasized", children: displayValue }),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { textStyle: "small", children: tooltipTimeValues.tooltipFormattedTimeValueStart }),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { textStyle: "small", children: tooltipTimeValues.tooltipFormattedTimeValueEnd })
          ]
        }),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Flex.Flex, { gap: 0, flexDirection: "column", children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { textStyle: "small", children: timeframeDuration }) })
      ] }),
      disabled: disabled || isInvalid,
      onOpenChange: onOpenTooltip,
      children
    }
  );
};
TimeframeSelectorTooltip.displayName = "TimeframeSelectorTooltip";
