// @dynatrace/strato-components/charts/core/components/legend/categorical/components/ChartLegendItemLabel/ChartLegendItemLabel.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ChartLegendItemLabel_exports = {};
__export(ChartLegendItemLabel_exports, {
  ChartLegendItemLabel: () => ChartLegendItemLabel
});
module.exports = __toCommonJS(ChartLegendItemLabel_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_Tooltip = require("../../../../../../../overlays/tooltip/Tooltip.js");
var import_constants = require("../../../../../constants.js");
var import_useChartLayoutRatio = require("../../../../chart-layout/hooks/useChartLayoutRatio.js");
var import_chars_font_width_map = require("../../../../text-ellipsis/chars-font-width-map.js");
var import_ChartTextEllipsis = require("../../../../text-ellipsis/ChartTextEllipsis.js");
var import_truncate_string = require("../../../../text-ellipsis/truncate-string.js");
var import_constants2 = require("../../constants.js");
var import_LegendGridSettingsContext = require("../../context/LegendGridSettingsContext.js");
var import_useLegend = require("../../hooks/useLegend.js");
var import_ChartLegendItem_css = require("../ChartLegendItem/ChartLegendItem.sty.js");
const MAX_WIDTH_GAP = 15;
const ChartLegendItemLabel = (props) => {
  const {
    item,
    buttonRef,
    visible,
    onButtonKeydown,
    onButtonClick,
    onMouseEnter,
    onMouseLeave
  } = props;
  const { name } = item;
  const textRef = (0, import_react.useRef)(null);
  const { readyForAutoRatio, ratio } = (0, import_useChartLayoutRatio.useChartLayoutRatio)();
  const { interactionsDisabled, truncationMode } = (0, import_useLegend.useLegend)();
  const { maxItemWidth } = (0, import_LegendGridSettingsContext.useLegendGridSettings)();
  let applyTruncation = false;
  if (buttonRef.current) {
    const textWidth = (0, import_truncate_string.calculateWidth)(
      item.name,
      import_chars_font_width_map.charsFontWidth12pixelsSize500Map
    );
    const availableWidth = buttonRef.current.clientWidth;
    applyTruncation = textWidth >= availableWidth;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Tooltip.Tooltip, { text: name, disabled: !applyTruncation, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "button",
    {
      "data-testid": "chart-legend-item-name",
      "data-dt-children-name": "LegendIsolateItem",
      ref: buttonRef,
      onClick: (evt) => onButtonClick(evt.nativeEvent),
      onKeyDown: (evt) => onButtonKeydown(evt.nativeEvent),
      onMouseEnter,
      onMouseLeave,
      tabIndex: interactionsDisabled ? -1 : 0,
      className: (0, import_clsx.default)(
        import_ChartLegendItem_css.baseLegendButton,
        import_ChartLegendItem_css.legendNameButtonOverride,
        (0, import_ChartLegendItem_css.legendNameButton)({
          isInteractive: !interactionsDisabled,
          isSubdued: !visible
        })
      ),
      style: {
        flexBasis: maxItemWidth + MAX_WIDTH_GAP
      },
      children: ratio !== "auto" || readyForAutoRatio ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_ChartTextEllipsis.ChartTextEllipsis,
        {
          truncationMode: truncationMode ?? import_constants.DEFAULT_TRUNCATION_MODE,
          "aria-label": "Series name",
          ref: textRef,
          charsFontWidthMap: import_chars_font_width_map.charsFontWidth12pixelsSize500Map,
          children: item.name
        }
      ) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { className: import_constants2.LABEL_ITEM_CLASSNAME, children: item.name })
    }
  ) });
};
