// @dynatrace/strato-geo/map/components/DefaultTooltip/DefaultTooltip.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DefaultTooltip_exports = {};
__export(DefaultTooltip_exports, {
  DefaultTooltip: () => DefaultTooltip
});
module.exports = __toCommonJS(DefaultTooltip_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_charts = require("@dynatrace/strato-components/charts");
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
var import_strato_icons = require("@dynatrace/strato-icons");
var import_use_truncation_mode = require("../../hooks/use-truncation-mode.js");
var import_tooltip_type_guards = require("../../utils/tooltip-type-guards.js");
const isChoroplethTooltip = (props) => (0, import_tooltip_type_guards.isRegion)(props.closestPoint);
const isConnectionTooltip = (props) => (0, import_tooltip_type_guards.isConnection)(props.closestPoint);
const isBubbleTooltip = (props) => !(0, import_lodash_es.isNil)(props.layerData);
const isDotTooltip = (props) => true;
const DefaultTooltip = (props) => {
  const defaultConfig = (0, import_charts._useDefaultConfig)();
  const getLabel = (0, import_charts._useDefaultLabels)();
  const truncationMode = (0, import_use_truncation_mode.useTruncationMode)();
  let locations = [];
  if (isChoroplethTooltip(props)) {
    const { closestPoint, seriesActions } = props;
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_charts._OverlayTooltip, { truncation: truncationMode, children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_charts._OverlayTooltip.SeriesHeader, { valueLabel: "" }),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_charts._OverlayTooltip.SeriesItem,
        {
          color: closestPoint.color,
          name: closestPoint.name,
          value: closestPoint.data.value?.toString() ?? "",
          actions: seriesActions ? () => seriesActions(closestPoint) : void 0
        }
      )
    ] });
  } else if (isConnectionTooltip(props)) {
    locations = props.closestPoint.data.path;
  } else if (isBubbleTooltip(props)) {
    const { closestPoint, layerData } = props;
    locations = layerData === void 0 || layerData === null || layerData.length === 0 ? [closestPoint.data] : layerData.map((point) => point.data);
  } else if (isDotTooltip(props)) {
    locations = [props.closestPoint.data];
  }
  const seriesLabel = getLabel?.("seriesLabel") ?? "";
  const variant = defaultConfig?.sharedVariant ? "shared" : "single";
  const seriesData = locations.map((location) => ({
    name: `${location.latitude}, ${location.longitude}`,
    value: ""
  }));
  const getActionsForItem = (current) => {
    if (props.actionsForItem) {
      return props.actionsForItem(current);
    }
    const { seriesActions: actions } = props;
    if (!actions || !locations.length) {
      return void 0;
    }
    const location = locations.find(
      (l) => `${l.latitude}, ${l.longitude}` === current.name
    );
    return location ? () => actions(location) : void 0;
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_charts._OverlayTooltip, { truncation: truncationMode, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_charts._OverlayTooltip.Header, { header1: seriesLabel }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_charts._OverlayTooltip.SeriesHeader, { disabled: true }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_charts._OverlayTooltip.Series,
      {
        activeItem: 0,
        data: seriesData,
        variant,
        seriesItemTemplate: (current) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_charts._OverlayTooltip.SeriesItem,
          {
            ...current,
            iconStyle: "large",
            actions: getActionsForItem(current),
            icon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_strato_icons.LocationMarkerFilledIcon,
              {
                style: {
                  fill: import_colors.default.Text.Neutral.Default
                }
              }
            )
          }
        )
      }
    )
  ] });
};
DefaultTooltip["displayName"] = "DefaultTooltip";
