// @dynatrace/strato-geo/esm/map/components/ChoroplethLayer/ChoroplethLayer.js
import { jsx, jsxs } from "react/jsx-runtime";
import { Layer, Source } from "@vis.gl/react-maplibre";
import { _useBehavioralTrackingProps as useBehavioralTrackingProps } from "@dynatrace/strato-components/core";
import { ChoroplethLayerTooltip } from "./ChoroplethLayerTooltip.js";
import { ChoroplethOutlineLayer } from "./ChoroplethOutlineLayer.js";
import { parseRegionDataToGeoJSON } from "./utils/parse-region-data-to-geo-json.js";
import {
  DEFAULT_CHOROPLETH_COLOR,
  SHAPE_OPACITY_DIMMED,
  SHAPE_OPACITY_DEFAULT
} from "../../constants.js";
import { useLayerColoringStrategy } from "../../hooks/use-layer-coloring-strategy.js";
import { useLegendInteractionHighlighting } from "../../hooks/use-legend-interaction-highlighting.js";
import { useLegendInteractionVisibility } from "../../hooks/use-legend-interaction-visibility.js";
import { useMapBaseLayerFeatures } from "../../hooks/use-map-base-layer-features.js";
import { useResolveChoroplethColor } from "../../hooks/use-resolve-color.js";
import { useTooltipTemplate } from "../../hooks/use-tooltip-template.js";
import { ChoroplethLayerTooltip as ChoroplethLayerTooltipSlot } from "../../slots/ChoroplethLayerTooltip.js";
import { isChoroplethTooltipTemplate } from "../../utils/tooltip-type-guards.js";
import { TooltipWrapper } from "../TooltipWrapper.js";
const ChoroplethLayer = (props) => {
  const { layerId, data, regionAccessor, children } = props;
  const [behavioralTrackingProps] = useBehavioralTrackingProps(props);
  const mapSource = useMapBaseLayerFeatures();
  const colorParser = useLayerColoringStrategy();
  const choroplethColor = useResolveChoroplethColor(
    DEFAULT_CHOROPLETH_COLOR,
    props,
    colorParser
  );
  const sourceId = layerId.replace("layer", "source");
  useLegendInteractionHighlighting(props, sourceId);
  useLegendInteractionVisibility(props, layerId);
  const dataGeoJSON = mapSource && parseRegionDataToGeoJSON(mapSource, data, regionAccessor, choroplethColor);
  const {
    handler: tooltipTemplate,
    hidden: tooltipHidden,
    symbolAlignment
  } = useTooltipTemplate(children, ChoroplethLayerTooltipSlot, {
    extractProps: true
  });
  return dataGeoJSON && /* @__PURE__ */ jsxs(Source, { id: sourceId, type: "geojson", data: dataGeoJSON, generateId: true, children: [
    /* @__PURE__ */ jsx(
      Layer,
      {
        id: layerId,
        source: sourceId,
        type: "fill",
        paint: {
          "fill-color": ["get", "__color"],
          "fill-opacity": [
            "case",
            ["==", ["feature-state", "legendHover"], false],
            SHAPE_OPACITY_DIMMED,
            ["==", ["feature-state", "active"], true],
            SHAPE_OPACITY_DEFAULT,
            ["==", ["feature-state", "hover"], true],
            SHAPE_OPACITY_DEFAULT,
            ["==", ["feature-state", "legendHover"], true],
            SHAPE_OPACITY_DEFAULT,
            ["==", ["feature-state", "isAnyActive"], true],
            SHAPE_OPACITY_DIMMED,
            SHAPE_OPACITY_DEFAULT
          ]
        },
        ...behavioralTrackingProps
      }
    ),
    /* @__PURE__ */ jsx(ChoroplethOutlineLayer, { sourceId }),
    /* @__PURE__ */ jsx(TooltipWrapper, { layerId, children: /* @__PURE__ */ jsx(
      ChoroplethLayerTooltip,
      {
        tooltipTemplate: isChoroplethTooltipTemplate(tooltipTemplate) ? tooltipTemplate : void 0,
        hidden: tooltipHidden,
        symbolAlignment
      }
    ) })
  ] });
};
ChoroplethLayer["displayName"] = "ChoroplethLayer";
export {
  ChoroplethLayer
};
