// @dynatrace/strato-components/charts/core/components/legend/range/components/RangeLegendMarkers/RangeLegendMarkers.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var RangeLegendMarkers_exports = {};
__export(RangeLegendMarkers_exports, {
  RangeLegendMarkers: () => RangeLegendMarkers
});
module.exports = __toCommonJS(RangeLegendMarkers_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_RangeLegendMarkers_css = require("./RangeLegendMarkers.sty.js");
var import_constants = require("../../../../../constants.js");
var import_constants2 = require("../../constants.js");
var import_RangeLegendOrientationContext = require("../../context/RangeLegendOrientationContext.js");
var import_RangeLegendScaleContext = require("../../context/RangeLegendScaleContext.js");
var import_use_legend_formatted_labels = require("../../hooks/use-legend-formatted-labels.js");
var import_get_label_alignment = require("../../utils/get-label-alignment.js");
var import_get_label_end_position = require("../../utils/get-label-end-position.js");
var import_get_markers = require("../../utils/get-markers.js");
var import_RangeLegendBoundaryLabels = require("../RangeLegendBoundaryLabels.js");
var import_RangeLegendItemLabel = require("../RangeLegendItemLabel/RangeLegendItemLabel.js");
var import_RangeLegendItemMarker = require("../RangeLegendItemMarker/RangeLegendItemMarker.js");
var import_RangeLegendSegmentsOverlay = require("../RangeLegendSegmentsOverlay.js");
const RangeLegendMarkers = (props) => {
  const { items, maxValue, minValue } = props;
  const hasMinValue = (0, import_lodash_es.isNumber)(minValue);
  const hasMaxValue = (0, import_lodash_es.isNumber)(maxValue);
  const { orientation } = (0, import_RangeLegendOrientationContext.useOrientation)();
  const { scale } = (0, import_RangeLegendScaleContext.useScale)();
  const [rangeStartPosition, rangeEndPosition] = scale.range();
  const isVertical = orientation === import_constants.ORIENTATION.Vertical;
  const { itemMarkers, singlePointRangeMarkers } = (0, import_react.useMemo)(
    () => (0, import_get_markers.getMarkers)(items, isVertical),
    [isVertical, items]
  );
  const internalRangeItems = (0, import_use_legend_formatted_labels.useLegendFormattedLabels)(items);
  const addedMarkers = /* @__PURE__ */ new Set([]);
  const markerLabel = isVertical ? `Labels from ${Array.from(itemMarkers)[itemMarkers.size - 1]} to ${Array.from(itemMarkers)[0]}` : `Labels from ${Array.from(itemMarkers)[0]} to ${Array.from(itemMarkers)[itemMarkers.size - 1]}`;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    "svg",
    {
      className: (0, import_RangeLegendMarkers_css.legendMarkersSvg)({
        isHorizontal: orientation === import_constants.ORIENTATION.Horizontal
      }),
      role: "group",
      "aria-label": "Chart legend",
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("g", { role: "list", "aria-label": markerLabel, children: [
          internalRangeItems.map((item) => {
            const hasFromLabel = itemMarkers.has(item.from) && !addedMarkers.has(item.from);
            const hasToLabel = itemMarkers.has(item.to) && !addedMarkers.has(item.to);
            if (!hasFromLabel && !hasToLabel) {
              return null;
            }
            addedMarkers.add(item.from);
            addedMarkers.add(item.to);
            const startPosition = scale(item.from);
            const endPosition = scale(item.to);
            const isSinglePointRange = startPosition === endPosition;
            const hasSameStartPosition = isVertical ? startPosition === rangeEndPosition : startPosition === rangeStartPosition;
            const isFirstItem = hasSameStartPosition && !hasMinValue;
            const hasSameEndPosition = isVertical ? endPosition === rangeStartPosition : endPosition === rangeEndPosition;
            const isLastItem = hasSameEndPosition && !hasMaxValue;
            const labelAlignment = (0, import_get_label_alignment.getLabelAlignment)(
              hasMaxValue,
              isLastItem,
              isVertical
            );
            const labelEndPosition = (0, import_get_label_end_position.getLabelEndPosition)(
              endPosition,
              isSinglePointRange,
              isLastItem
            );
            return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_react.Fragment, { children: [
              hasFromLabel && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_RangeLegendItemLabel.RangeLegendItemLabel,
                {
                  labelId: item.from,
                  startPosition,
                  endPosition: startPosition,
                  text: item.labels.from,
                  alignment: hasMinValue || !isFirstItem ? import_constants2.LABEL_ALIGNMENT.Center : import_constants2.LABEL_ALIGNMENT.End
                }
              ),
              hasToLabel && !isSinglePointRange && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_RangeLegendItemLabel.RangeLegendItemLabel,
                {
                  labelId: item.to,
                  startPosition: labelEndPosition,
                  endPosition: labelEndPosition,
                  text: item.labels.to,
                  alignment: labelAlignment
                }
              )
            ] }, `label-'${item.from}`);
          }),
          Array.from(itemMarkers).map((marker) => {
            const position = scale(marker);
            return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_RangeLegendItemMarker.RangeLegendItemMarker,
              {
                position,
                markerValue: marker,
                isReversed: isVertical,
                isSinglePointRange: singlePointRangeMarkers.has(marker)
              },
              `marker-'${marker}`
            );
          }),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_RangeLegendBoundaryLabels.RangeLegendBoundaryLabels,
            {
              items: internalRangeItems,
              maxValue,
              minValue
            }
          )
        ] }),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_RangeLegendSegmentsOverlay.RangeLegendSegmentsOverlay, { itemMarkers: Array.from(itemMarkers) })
      ]
    }
  );
};
