// @dynatrace/strato-components/core/hooks/intent.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var intent_exports = {};
__export(intent_exports, {
  useIntent: () => useIntent,
  useIntentWithResponse: () => useIntentWithResponse
});
module.exports = __toCommonJS(intent_exports);
var import_react = require("react");
var import_navigation = require("@dynatrace-sdk/navigation");
function getAppIconURL(appId) {
  if (!appId) {
    return null;
  }
  return `/platform/app-engine/registry/v1/app-icons/${appId}`;
}
function useIntent({ payload, options }) {
  let iconURL = null;
  if (options?.recommendedAppId) {
    iconURL = getAppIconURL(options.recommendedAppId);
  }
  function send() {
    if (!options) {
      return (0, import_navigation.sendIntent)(payload);
    }
    return (0, import_navigation.sendIntent)(payload, options);
  }
  return { send, iconURL };
}
function reducer(_, action) {
  switch (action.type) {
    case "send":
      return { status: "pending" };
    case "success":
      return { status: "success", data: action.payload };
    case "error":
      return { status: "error", error: action.payload };
    default:
      throw new Error();
  }
}
function useIntentWithResponse({
  payload,
  options
}) {
  const iconURL = getAppIconURL(options.recommendedAppId);
  const [state, dispatch] = (0, import_react.useReducer)(reducer, {
    status: "idle"
  });
  async function send(overridePayload) {
    dispatch({ type: "send" });
    const res = await (0, import_navigation.sendIntentWithResponse)(
      overridePayload || payload,
      options
    );
    if (!res) {
      const error = new Error("No response received");
      dispatch({
        type: "error",
        payload: error
      });
      return {
        status: "error",
        error
      };
    }
    if (res.ok) {
      dispatch({ type: "success", payload: { response: res?.data } });
      return { status: "success", data: { response: res?.data } };
    } else {
      const error = new Error(`Intent Error: ${res.error.reason}`);
      dispatch({
        type: "error",
        payload: error
      });
      return {
        status: "error",
        error
      };
    }
  }
  return { send, state, iconURL };
}
