// @dynatrace/strato-components/core/hooks/useOverlayStacking.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useOverlayStacking_exports = {};
__export(useOverlayStacking_exports, {
  useOverlayStacking: () => useOverlayStacking
});
module.exports = __toCommonJS(useOverlayStacking_exports);
var import_react = require("react");
var import_OverlayStacking = require("../contexts/OverlayStacking.js");
const ElevationMapping = {
  default: 1,
  dropdown: 10,
  sticky: 20,
  overlay: 30,
  "modal-backdrop": 40,
  "feature-highlight": 50,
  microguide: 60
};
function useOverlayStacking(topOffset = 0, additionalElevationToken) {
  const overlayStacking = (0, import_react.useContext)(import_OverlayStacking.OverlayStackingContext);
  const additionalElevation = additionalElevationToken ? ElevationMapping[additionalElevationToken] : void 0;
  const overlayContainerProps = (0, import_react.useMemo)(() => {
    return {
      style: {
        position: "fixed",
        top: topOffset,
        left: 0,
        zIndex: overlayStacking > 0 ? `calc(${overlayStacking * 100 + (additionalElevation ?? 0)})` : additionalElevation
      }
    };
  }, [topOffset, overlayStacking, additionalElevation]);
  return overlayContainerProps;
}
