// @dynatrace/strato-components/core/hooks/usePermissions.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var usePermissions_exports = {};
__export(usePermissions_exports, {
  usePermissions: () => usePermissions
});
module.exports = __toCommonJS(usePermissions_exports);
var import_react_query = require("@tanstack/react-query");
var import_client_platform_management_service = require("@dynatrace-sdk/client-platform-management-service");
var import_permissionsQueryClient = require("../utils/permissionsQueryClient.js");
async function resolvePermissions(permissions) {
  const result = await import_client_platform_management_service.effectivePermissionsClient.resolveEffectivePermissions({
    body: {
      permissions: permissions.map((permission) => ({ permission }))
    }
  });
  return (result ?? []).map(({ permission, granted }) => ({
    permission,
    granted
  }));
}
function usePermissions(permissions) {
  const queryKey = [
    "permissions",
    ...permissions.slice().sort((a, b) => a.localeCompare(b))
  ];
  const { data, isLoading, isError } = (0, import_react_query.useQuery)(
    {
      queryKey,
      queryFn: () => resolvePermissions(permissions),
      enabled: permissions.length > 0
    },
    import_permissionsQueryClient.permissionsQueryClient
  );
  const details = data ?? [];
  const isAllowed = !isLoading && !isError && (permissions.length === 0 || details.length > 0 && details.every(
    ({ granted }) => granted === "true" || granted === "condition"
  ));
  return { isAllowed, isLoading, details };
}
