// @dynatrace/strato-components/charts/xy-chart/hooks/useSortRectangleDatapoints.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useSortRectangleDatapoints_exports = {};
__export(useSortRectangleDatapoints_exports, {
  isValidNumeric: () => isValidNumeric,
  useSortRectangleDatapoints: () => useSortRectangleDatapoints
});
module.exports = __toCommonJS(useSortRectangleDatapoints_exports);
var import_lodash_es = require("lodash");
var import_detect_scale = require("../../core/utils/detect-scale.js");
var import_is_rectangle_selected = require("../components/overlay/utils/is-rectangle-selected.js");
var import_get_scale = require("../utils/get-scale.js");
var import_value_representation = require("../utils/value-representation.js");
const toNumber = (value) => typeof value === "number" ? value : value.getTime();
const parseValue = (value) => {
  if (typeof value === "number" || typeof value === "string") {
    return value;
  }
  return value.getTime();
};
const hashColumnKey = (x0, x1) => x1 && x0 < x1 ? `${parseValue(x0)}_${parseValue(x1 ?? x0)}` : `${parseValue(x1 ?? x0)}_${parseValue(x0)}`;
const sortRangeKeys = (keys) => keys.toSorted((a, b) => {
  const [a0] = a.split("_");
  const [b0] = b.split("_");
  if (!isNaN(parseInt(a0)) && !isNaN(parseInt(b0))) {
    return parseInt(a0) - parseInt(b0);
  }
  return 0;
});
const isValidNumeric = (v) => v !== void 0 && typeof v !== "string";
const applySerialization = (range, columnsMap, callback, rectsByRange) => {
  for (const columnKey of range) {
    const rectsInColumn = columnsMap.get(columnKey);
    if (rectsInColumn) {
      for (const rect of rectsInColumn) {
        const mapped = callback(rect);
        if (mapped) {
          rectsByRange.push(mapped);
        }
      }
    }
  }
};
const useSortRectangleDatapoints = (seriesVariantsProps) => {
  const rectAxisSelectConfig = /* @__PURE__ */ new Map();
  function updateBoundariesForAxis(boundaries, axisName, v0, v1) {
    if (typeof v0 === "string" || typeof v1 === "string") {
      return;
    }
    if (isValidNumeric(v0) && isValidNumeric(v1)) {
      const v0Num = toNumber(v0);
      const v1Num = toNumber(v1);
      if (boundaries[axisName] && (0, import_lodash_es.isNumber)(boundaries[axisName].min) && (0, import_lodash_es.isNumber)(boundaries[axisName].max)) {
        boundaries[axisName].min = Math.min(
          toNumber(boundaries[axisName].min),
          v0Num,
          v1Num
        );
        boundaries[axisName].max = Math.max(
          toNumber(boundaries[axisName].max),
          v0Num,
          v1Num
        );
      } else {
        boundaries[axisName] = {
          min: Math.min(v0Num, v1Num),
          max: Math.max(v0Num, v1Num)
        };
      }
    }
  }
  const handleAddValue = (rect) => {
    const { x0, x1, y0, y1 } = rect.datapoint;
    const xAxisId = rect.series.xAxisId;
    const yAxisId = rect.series.yAxisId;
    if ((0, import_lodash_es.isUndefined)(x0) && (0, import_lodash_es.isUndefined)(x1) && (0, import_lodash_es.isUndefined)(y0) && (0, import_lodash_es.isUndefined)(y1)) {
      return;
    }
    const isVerticalBar = typeof x0 === "string" || typeof y0 !== "string" && !(0, import_lodash_es.isUndefined)(x0);
    const isHorizontalBar = typeof y0 === "string" || typeof x0 !== "string" && (0, import_lodash_es.isUndefined)(x0) && !(0, import_lodash_es.isUndefined)(y0);
    const primaryAxisId = isVerticalBar ? xAxisId : isHorizontalBar ? yAxisId : xAxisId;
    const secondaryAxisId = isVerticalBar ? yAxisId : isHorizontalBar ? xAxisId : yAxisId;
    const primaryV0 = isVerticalBar ? x0 : isHorizontalBar ? y0 : x0;
    const primaryV1 = isVerticalBar ? x1 : isHorizontalBar ? y1 : x1;
    const secondaryV0 = isVerticalBar ? y0 : isHorizontalBar ? x0 : y0;
    const secondaryV1 = isVerticalBar ? y1 : isHorizontalBar ? x1 : y1;
    if (!rectAxisSelectConfig.has(primaryAxisId)) {
      rectAxisSelectConfig.set(primaryAxisId, {
        uniqueColumns: /* @__PURE__ */ new Set(),
        uniqueRows: /* @__PURE__ */ new Set(),
        columnsMap: /* @__PURE__ */ new Map(),
        boundaries: {}
      });
    }
    const axisConfig = rectAxisSelectConfig.get(primaryAxisId) ?? {
      uniqueColumns: /* @__PURE__ */ new Set(),
      uniqueRows: /* @__PURE__ */ new Set(),
      columnsMap: /* @__PURE__ */ new Map(),
      boundaries: {}
    };
    let hashedKey;
    if (typeof primaryV0 === "string" || typeof primaryV1 === "string") {
      hashedKey = primaryV0;
    } else if (primaryV0 !== void 0) {
      hashedKey = hashColumnKey(primaryV0, primaryV1);
      if (secondaryV0 !== void 0) {
        updateBoundariesForAxis(
          axisConfig.boundaries,
          secondaryAxisId,
          secondaryV0,
          secondaryV1
        );
        const rowKey = hashColumnKey(secondaryV0, secondaryV1);
        axisConfig.uniqueRows.add(rowKey);
      }
    }
    if (isValidNumeric(primaryV0) && primaryV0 !== void 0) {
      updateBoundariesForAxis(
        axisConfig.boundaries,
        primaryAxisId,
        primaryV0,
        primaryV1
      );
    }
    if (!hashedKey) {
      return;
    }
    if (!axisConfig.uniqueColumns.has(hashedKey)) {
      axisConfig.uniqueColumns.add(hashedKey);
      axisConfig.columnsMap.set(hashedKey, []);
    }
    const rectsInRange = axisConfig.columnsMap.get(hashedKey) || [];
    rectsInRange.push(rect);
    axisConfig.columnsMap.set(hashedKey, rectsInRange);
  };
  const getAxisSortedValues = (axisId) => {
    const sortedData = rectAxisSelectConfig.get(axisId);
    if (sortedData) {
      const { uniqueColumns, uniqueRows, columnsMap } = sortedData;
      const sortedColumns = sortRangeKeys(Array.from(uniqueColumns));
      const sortedRows = sortRangeKeys(Array.from(uniqueRows));
      return {
        sortedColumns,
        columnsMap,
        sortedRows
      };
    }
    return void 0;
  };
  const getAxisValues = (axisId) => rectAxisSelectConfig.get(axisId);
  const handleSearchRange = (start, end, sortedColumns) => {
    const sX = start instanceof Date ? start.getTime() : Number(start);
    const eX = end instanceof Date ? end.getTime() : Number(end);
    const [min, max] = sX <= eX ? [sX, eX] : [eX, sX];
    return sortedColumns.filter((key) => {
      const [x0Str, x1Str] = key.split("_");
      const colX0 = Number(x0Str);
      const colX1 = Number(x1Str);
      const isInRange = (0, import_is_rectangle_selected.isRectangleSelected)({
        x: {
          rectangleStart: colX0,
          rectangleEnd: colX1,
          selectionStart: min,
          selectionEnd: max
        }
      });
      return isInRange;
    });
  };
  const handleGetRectanglesByNumericalRangeForHorizontalBar = (start, end, axisConfig, callback) => {
    const rectsByRange = [];
    const [selectionStart, selectionEnd] = start < end ? [start, end] : [end, start];
    for (const rectsInColumn of axisConfig.columnsMap.values()) {
      for (const rect of rectsInColumn) {
        const rectanglePoint = callback(rect);
        if (!rectanglePoint) {
          continue;
        }
        const barStart = rectanglePoint.x;
        const barEnd = rectanglePoint.x + rectanglePoint.width;
        const isInRange = (0, import_is_rectangle_selected.isRectangleSelected)({
          x: {
            rectangleStart: barStart,
            rectangleEnd: barEnd,
            selectionStart,
            selectionEnd
          }
        });
        if (isInRange) {
          rectsByRange.push(rectanglePoint);
        }
      }
    }
    return rectsByRange;
  };
  const handleGetRectanglesByNumericalRangeForVerticalBar = (start, end, callback, axisId) => {
    const sortedValuesByAxis = getAxisSortedValues(axisId);
    const rectsByRange = [];
    if (sortedValuesByAxis) {
      const { sortedColumns, columnsMap } = sortedValuesByAxis;
      const range = handleSearchRange(start, end, sortedColumns);
      applySerialization(range, columnsMap, callback, rectsByRange);
    }
    return rectsByRange;
  };
  const handleGetRectanglesByCategoricalRange = (start, end, axisId, callback) => {
    const categoricalValues = getAxisValues(axisId);
    const rectsByRange = [];
    if (categoricalValues) {
      const { columnsMap } = categoricalValues;
      const [selectionStart, selectionEnd] = start < end ? [start, end] : [end, start];
      for (const [, rectsInColumn] of columnsMap) {
        for (const rect of rectsInColumn) {
          const rectanglePoint = callback(rect);
          const barStart = rectanglePoint.x;
          const barEnd = rectanglePoint.x + rectanglePoint.width;
          const isInRange = (0, import_is_rectangle_selected.isRectangleSelected)({
            x: {
              rectangleStart: barStart,
              rectangleEnd: barEnd,
              selectionStart,
              selectionEnd
            }
          });
          if (isInRange) {
            rectsByRange.push(rectanglePoint);
          }
        }
      }
    }
    return rectsByRange;
  };
  const isHorizontalBarConfiguration = (axisConfig) => {
    if (axisConfig.columnsMap.size === 0) {
      return false;
    }
    const firstColumn = Array.from(axisConfig.columnsMap.values())[0];
    const firstDatapoint = firstColumn?.[0]?.datapoint;
    return firstDatapoint && (0, import_lodash_es.isUndefined)(firstDatapoint.x0);
  };
  const searchAllAxisDatapoints = (start, end, callback, axesIds, axisIdToScales, axisIdToScaleDomain, axisIDToIsRelative) => {
    const axesCurrentSelection = {};
    const highlightedDatapoints = [];
    if (!axisIdToScales) {
      return { axesCurrentSelection, highlightedDatapoints };
    }
    const axesIdsArray = typeof axesIds === "string" ? [axesIds] : axesIds;
    axesIdsArray.forEach((axisId) => {
      if (rectAxisSelectConfig.has(axisId)) {
        const currentScale = (0, import_get_scale.getCurrentScale)(axisIdToScales, axisId);
        if ((0, import_detect_scale.isBandScale)(currentScale)) {
          const rectsByRange = handleGetRectanglesByCategoricalRange(
            start,
            end,
            axisId,
            callback
          );
          for (const rect of rectsByRange) {
            highlightedDatapoints.push(rect);
          }
        } else {
          const axisConfig = rectAxisSelectConfig.get(axisId);
          if (axisConfig && isHorizontalBarConfiguration(axisConfig)) {
            const horizontalBarResults = handleGetRectanglesByNumericalRangeForHorizontalBar(
              start,
              end,
              axisConfig,
              callback
            );
            highlightedDatapoints.push(...horizontalBarResults);
            return;
          }
          let startDomain = currentScale.invert(start);
          let endDomain = currentScale.invert(Math.ceil(end));
          if (axisIDToIsRelative.get(axisId)) {
            const domain = axisIdToScaleDomain.get(axisId);
            startDomain = (0, import_value_representation.toAbsolute)(startDomain, domain);
            endDomain = (0, import_value_representation.toAbsolute)(endDomain, domain);
          }
          const seriesHighlightedDatapoints = handleGetRectanglesByNumericalRangeForVerticalBar(
            startDomain,
            endDomain,
            callback,
            axisId
          );
          for (const datapoint of seriesHighlightedDatapoints) {
            highlightedDatapoints.push(datapoint);
          }
          axesCurrentSelection[axisId] = seriesHighlightedDatapoints;
        }
      }
    });
    return { axesCurrentSelection, highlightedDatapoints };
  };
  return {
    addValue: handleAddValue,
    getCurrentSelectConfig: () => rectAxisSelectConfig,
    searchAllAxisDatapoints
  };
};
