// @dynatrace/strato-components/charts/xy-chart/hooks/useXYChartDataProvider.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useXYChartDataProvider_exports = {};
__export(useXYChartDataProvider_exports, {
  useXYChartDataProvider: () => useXYChartDataProvider
});
module.exports = __toCommonJS(useXYChartDataProvider_exports);
var import_react_intl = require("react-intl");
var import_useAxesConfig = require("./useAxesConfig.js");
var import_useColorPalette = require("./useColorPalette.js");
var import_useSortLinearDatapoints = require("./useSortLinearDatapoints.js");
var import_useSortRectangleDatapoints = require("./useSortRectangleDatapoints.js");
var import_factory = require("../processing-strategies/factory/factory.js");
var import_area_series = require("../processing-strategies/strategies/area-series.js");
var import_bar_series = require("../processing-strategies/strategies/bar-series.js");
var import_dot_series = require("../processing-strategies/strategies/dot-series.js");
var import_line_series = require("../processing-strategies/strategies/line-series.js");
var import_rect_series = require("../processing-strategies/strategies/rect-series.js");
var import_get_datapoints = require("../processing-strategies/utils/access-datapoint/get-datapoints.js");
var import_selectors = require("../store/selectors.js");
var import_guards = require("../types/guards.js");
var import_axis_interactions = require("../utils/axis-interactions.js");
var import_combine_data_with_corresponding_series = require("../utils/combine-data-with-corresponding-series.js");
var import_combine_series_with_axes = require("../utils/combine-series-with-axes.js");
function useXYChartDataProvider(seriesVariantsProps, data) {
  const axes = (0, import_useAxesConfig.useAxesConfig)();
  const colorPalette = (0, import_useColorPalette.useColorPalette)();
  const intl = (0, import_react_intl.useIntl)();
  const axisIdToScaleDomain = /* @__PURE__ */ new Map();
  const axisWithSeries = /* @__PURE__ */ new Set();
  const datapointMetadataMap = /* @__PURE__ */ new Map();
  const axisIdToCategoryDatum = /* @__PURE__ */ new Map();
  const {
    addValue,
    getCurrentSelectConfig: getCurrentShapeSelectConfig,
    searchAllAxisDatapoints: searchAllAxisShapeDatapoints
  } = (0, import_useSortLinearDatapoints.useSortLinearDatapoints)(seriesVariantsProps, datapointMetadataMap);
  const {
    addValue: addRectangleValue,
    searchAllAxisDatapoints: searchAllAxisRectangleDatapoints,
    getCurrentSelectConfig: getCurrentRectangleSelectConfig
  } = (0, import_useSortRectangleDatapoints.useSortRectangleDatapoints)(seriesVariantsProps);
  const seriesWithAxes = (0, import_combine_series_with_axes.combineSeriesWithAxes)(seriesVariantsProps, axes);
  const dataWithSeries = (0, import_combine_data_with_corresponding_series.combineDataWithCorrespondingSeries)(
    data,
    seriesWithAxes
  );
  const seriesWithDatapoints = [];
  const seriesToMetadata = /* @__PURE__ */ new Map();
  const seriesMap = /* @__PURE__ */ new Map();
  const seriesIndexes = [];
  const rectSeriesStrategy = new import_rect_series.RectSeriesStrategy({
    addValue: addRectangleValue
  });
  const dotSeriesStrategy = new import_dot_series.DotSeriesStrategy({ addValue });
  const lineSeriesStrategy = new import_line_series.LineSeriesStrategy({ addValue });
  const barSeriesStrategy = new import_bar_series.BarSeriesStrategy({
    addValue: addRectangleValue
  });
  const areaSeriesStrategy = new import_area_series.AreaSeriesStrategy({ addValue });
  const strategies = [
    rectSeriesStrategy,
    dotSeriesStrategy,
    lineSeriesStrategy,
    barSeriesStrategy,
    areaSeriesStrategy
  ];
  const filteredSeriesIds = (0, import_selectors.useFilteredSeriesIdsState)();
  const mutableContext = {
    seriesMap,
    seriesWithDatapoints,
    seriesToMetadata,
    axisIdToScaleDomain,
    seriesIndexes,
    intl,
    colorPalette,
    datapointMetadataMap,
    filteredSeriesIds,
    axisWithSeries
  };
  const captureCategoryDatum = (axis, accessor, datum) => {
    if (!(0, import_guards.isCategorical)(axis.type) || !accessor) {
      return;
    }
    const rawValue = (0, import_get_datapoints.getDatumValue)(datum, accessor);
    if (typeof rawValue !== "string" && typeof rawValue !== "number") {
      return;
    }
    const raw = String(rawValue);
    let perAxis = axisIdToCategoryDatum.get(axis.id);
    if (!perAxis) {
      perAxis = /* @__PURE__ */ new Map();
      axisIdToCategoryDatum.set(axis.id, perAxis);
    }
    if (!perAxis.has(raw)) {
      perAxis.set(raw, datum);
    }
  };
  for (const { data: data2, series } of dataWithSeries.filter(
    ({ series: series2 }) => series2.length
  )) {
    for (const { seriesProps, xAxis, yAxis } of series) {
      const strategy = (0, import_factory.getStrategyFactory)(strategies, seriesProps);
      if (strategy?.preprocess) {
        strategy.preprocess(data2, seriesProps, xAxis, yAxis, mutableContext);
      }
    }
    for (const datum of data2) {
      for (const { seriesProps, xAxis, yAxis } of series) {
        const strategy = (0, import_factory.getStrategyFactory)(strategies, seriesProps);
        if (strategy) {
          const processResult = strategy.process(
            datum,
            seriesProps,
            xAxis,
            yAxis,
            mutableContext
          );
          if (processResult === "continue") {
            continue;
          }
        } else {
          throw new Error(
            `No strategy found to process series with id: ${seriesProps.id}`
          );
        }
        captureCategoryDatum(xAxis, seriesProps.x0Accessor, datum);
        captureCategoryDatum(yAxis, seriesProps.y0Accessor, datum);
      }
    }
  }
  const validAxes = axes.filter(
    (axis) => axisIdToScaleDomain.has(axis.id) && axisWithSeries.has(axis.id)
  );
  const linearSelectionContextValue = {
    selectionConfigByAxisId: getCurrentShapeSelectConfig(),
    searchAllAxisDatapoints: searchAllAxisShapeDatapoints
  };
  const rectSelectionContextValue = {
    selectionConfigByAxisId: getCurrentRectangleSelectConfig(),
    searchAllAxisDatapoints: searchAllAxisRectangleDatapoints
  };
  return {
    axesMetadata: {
      isZoomEnabled: !!validAxes.find(import_axis_interactions.isZoomEnabledFor),
      isPanEnabled: !!validAxes.find(import_axis_interactions.isPanEnabledFor),
      isInfinite: !!validAxes.find(import_axis_interactions.isInfiniteZoomEnabledFor)
    },
    axes: validAxes,
    axisIdToScaleDomain,
    axisIdToCategoryDatum,
    seriesIdToValueScaleDomain: rectSeriesStrategy.seriesIdToValueScaleDomain,
    datapointMetadataMap,
    linearSelectionContextValue,
    rectSelectionContextValue,
    seriesWithDatapoints,
    filteredSeriesWithDatapoints: seriesWithDatapoints.filter((series) => {
      if (series.type === "rect") {
        series.filteredData = series.data.filter(
          (datapoint) => {
            return datapointMetadataMap.get(datapoint)?.isVisible !== false;
          }
        );
      }
      return seriesToMetadata.get(series)?.isVisible !== false;
    }),
    seriesToMetadata,
    axisIdToDimension: barSeriesStrategy.axisIdToDimension
  };
}
