// @dynatrace/strato-components/charts/core/components/interactions-overlay/utils/get-gestures.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var get_gestures_exports = {};
__export(get_gestures_exports, {
  getGestures: () => getGestures
});
module.exports = __toCommonJS(get_gestures_exports);
var import_reset = require("./reset.js");
const getGestures = ({
  lastMousePositionRef,
  panDeltaRef,
  onZoomStart,
  onZoomChange,
  onZoomEnd,
  onPanStart,
  onPanChange,
  onPanEnd
}) => ({
  onPinchStart: () => {
    (0, import_reset.reset)(lastMousePositionRef, panDeltaRef);
    onZoomStart?.({ triggeredBy: "pointer" });
  },
  onPinchMove: (event) => {
    const { delta } = event;
    const pinchIn = delta > 0;
    onZoomChange?.({
      triggeredBy: "pointer",
      deltaStart: delta,
      deltaEnd: -delta,
      behavior: pinchIn ? "zoom-in" : "zoom-out"
    });
  },
  onPinchEnd: () => {
    onZoomEnd?.({ triggeredBy: "pointer" });
  },
  onSwipeStart: () => {
    onPanStart?.({ triggeredBy: "pointer" });
  },
  onSwipeMove: (event) => {
    const { deltaX } = event;
    onPanChange?.({ triggeredBy: "pointer", delta: -deltaX });
  },
  onSwipeEnd: () => {
    onPanEnd?.({ triggeredBy: "pointer" });
  }
});
