// @dynatrace/strato-components/charts/core/components/interactions-overlay/utils/get-zoom-interactions.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var get_zoom_interactions_exports = {};
__export(get_zoom_interactions_exports, {
  getZoomInteractions: () => getZoomInteractions
});
module.exports = __toCommonJS(get_zoom_interactions_exports);
var import_reset = require("./reset.js");
var import_dragInteractions = require("../interactions/pan/dragInteractions.js");
var import_dragInteractions2 = require("../interactions/zoom/dragInteractions.js");
var import_keyboardInteractions = require("../interactions/zoom/keyboardInteractions.js");
var import_mouseInteractions = require("../interactions/zoom/mouseInteractions.js");
const getZoomInteractions = ({
  cursorMode,
  selectionArea,
  eventCancelled,
  onSelectEnd,
  lastMousePositionRef,
  pinned,
  onHover,
  onTogglePin,
  setCursorMode,
  height,
  onSelectChange,
  onSelectStart,
  onReset,
  width,
  onZoomChange,
  onPanChange,
  panDeltaRef,
  onPanStart,
  onPanEnd
}) => {
  return {
    onMouseHover: (event) => {
      if (selectionArea) {
        return;
      }
      (0, import_mouseInteractions.zoomMouseHover)({
        event,
        pinned,
        lastMousePositionRef,
        cursorMode,
        setCursorMode,
        onHover
      });
    },
    onMouseClickOrTap: (event) => (0, import_mouseInteractions.zoomMouseClickOrTap)({
      event,
      width,
      lastMousePositionRef,
      onZoomChange
    }),
    onDragStart: (event) => (0, import_dragInteractions2.zoomDragStart)({
      event,
      onSelectStart
    }),
    onDragMove: (event) => (0, import_dragInteractions2.zoomDragMove)({ event, height, width, onSelectChange }),
    onDragEnd: (event) => (0, import_dragInteractions2.zoomDragEnd)({
      event,
      height,
      width,
      lastMousePositionRef,
      onSelectEnd,
      onZoomChange
    }),
    onDoubleClickOrTap: (triggeredBy) => {
      onReset?.({
        triggeredBy
      });
      (0, import_reset.reset)(lastMousePositionRef);
    },
    onWheel: (event) => (0, import_mouseInteractions.zoomMouseWheel)({
      event,
      lastMousePositionRef,
      width,
      onZoomChange,
      onPanChange
    }),
    onKeyEscape: () => (0, import_keyboardInteractions.zoomKeyboardEscape)({
      selectionArea,
      eventCancelled,
      onSelectEnd,
      lastMousePositionRef,
      pinned,
      onHover,
      onTogglePin
    }),
    onKeyEnter: () => (0, import_keyboardInteractions.zoomKeyboardEnter)({
      selectionArea,
      lastMousePositionRef,
      width,
      onSelectEnd,
      onZoomChange,
      onHover
    }),
    onKeyRight: () => {
      return;
    },
    onKeyLeft: () => {
      return;
    },
    onKeyUp: (isTemporaryZoomMode) => {
      if (isTemporaryZoomMode) {
        (0, import_keyboardInteractions.zoomKeyboardUp)({ width, onZoomChange });
      }
    },
    onKeyDown: (isTemporaryZoomMode) => {
      if (isTemporaryZoomMode) {
        (0, import_keyboardInteractions.zoomKeyboardDown)({ width, onZoomChange });
      }
    },
    onDragStartWithMiddleButton: () => (0, import_dragInteractions.panDragStartWithMiddleButton)({ panDeltaRef, onPanStart }),
    onDragMoveWithMiddleButton: (event) => (0, import_dragInteractions.panDragMoveWithMiddleButton)({ event, panDeltaRef, onPanChange }),
    onDragEndWithMiddleButton: (event) => (0, import_dragInteractions.panDragEndWithMiddleButton)({
      event,
      panDeltaRef,
      lastMousePositionRef,
      onPanEnd
    })
  };
};
