// @dynatrace/strato-components/charts/core/components/interactions-overlay/utils/get-explore-interations.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var get_explore_interations_exports = {};
__export(get_explore_interations_exports, {
  getExploreInteractions: () => getExploreInteractions
});
module.exports = __toCommonJS(get_explore_interations_exports);
var import_dragInteractions = require("../interactions/explore/dragInteractions.js");
var import_keyboardInteractions = require("../interactions/explore/keyboardInteractions.js");
var import_mouseInteractions = require("../interactions/explore/mouseInteractions.js");
var import_dragInteractions2 = require("../interactions/pan/dragInteractions.js");
var import_keyboardInteractions2 = require("../interactions/zoom/keyboardInteractions.js");
var import_reset = require("./reset.js");
const getExploreInteractions = ({
  selectionArea,
  eventCancelled,
  onSelectEnd,
  lastMousePositionRef,
  pinned,
  onHover,
  onTogglePin,
  onModeChange,
  setCursorMode,
  height,
  onSelectChange,
  onSelectStart,
  onReset,
  width,
  onZoomChange,
  isZoomXDisabled,
  onPanChange,
  panDeltaRef,
  onPanStart,
  onPanEnd
}) => {
  return {
    onMouseHover: (event) => (0, import_mouseInteractions.exploreMouseHover)({
      event,
      pinned,
      selectionArea,
      lastMousePositionRef,
      setCursorMode,
      onHover
    }),
    onMouseClickOrTap: (event) => (0, import_mouseInteractions.exploreMouseClickOrTap)({
      event,
      selectionArea,
      onHover,
      onTogglePin,
      pinned,
      onSelectEnd,
      lastMousePositionRef
    }),
    onDragStart: (event) => (0, import_dragInteractions.exploreDragStart)({ event, onSelectStart }),
    onDragMove: (event) => (0, import_dragInteractions.exploreDragMove)({ event, height, width, onSelectChange }),
    onDragEnd: (event) => (0, import_dragInteractions.exploreDragEnd)({
      event,
      height,
      width,
      onSelectEnd,
      lastMousePositionRef
    }),
    onDoubleClickOrTap: (triggeredBy) => {
      const mousePosition = lastMousePositionRef.current;
      onReset?.({
        triggeredBy,
        mousePosition
      });
      (0, import_reset.reset)(lastMousePositionRef);
    },
    onWheel: (event) => {
      (0, import_mouseInteractions.exploreMouseWheel)({ event, onPanChange });
    },
    onKeyEscape: () => (0, import_keyboardInteractions.exploreKeyboardEscape)({
      selectionArea,
      eventCancelled,
      onSelectEnd,
      lastMousePositionRef,
      pinned,
      onHover,
      onTogglePin,
      onModeChange
    }),
    onKeyEnter: () => (0, import_keyboardInteractions.exploreKeyboardEnter)({
      selectionArea,
      lastMousePositionRef,
      width,
      onSelectEnd,
      onZoomChange,
      onHover,
      isZoomXDisabled
    }),
    onKeyRight: () => {
      return;
    },
    onKeyLeft: () => {
      return;
    },
    onKeyUp: (isTemporaryZoomMode) => {
      if (isTemporaryZoomMode) {
        (0, import_keyboardInteractions2.zoomKeyboardUp)({ width, onZoomChange });
      }
    },
    onKeyDown: (isTemporaryZoomMode) => {
      if (isTemporaryZoomMode) {
        (0, import_keyboardInteractions2.zoomKeyboardDown)({ width, onZoomChange });
      }
    },
    onDragStartWithMiddleButton: () => (0, import_dragInteractions2.panDragStartWithMiddleButton)({ panDeltaRef, onPanStart }),
    onDragMoveWithMiddleButton: (event) => (0, import_dragInteractions2.panDragMoveWithMiddleButton)({ event, panDeltaRef, onPanChange }),
    onDragEndWithMiddleButton: (event) => (0, import_dragInteractions2.panDragEndWithMiddleButton)({
      event,
      panDeltaRef,
      lastMousePositionRef,
      onPanEnd
    })
  };
};
