// @dynatrace/strato-components/charts/core/components/interactions-overlay/utils/get-interface-interactions.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var get_interface_interactions_exports = {};
__export(get_interface_interactions_exports, {
  getInterfaceInteractions: () => getInterfaceInteractions
});
module.exports = __toCommonJS(get_interface_interactions_exports);
var import_get_active_mode = require("./get-active-mode.js");
const getInterfaceInteractions = (interactions, mode, isZoomDisabled, isZoomXDisabled, isPanDisabled) => ({
  onDragStart: (event) => interactions[(0, import_get_active_mode.getActiveMode)(mode, event, isZoomXDisabled)].onDragStart(
    event
  ),
  onDragMove: (event) => interactions[(0, import_get_active_mode.getActiveMode)(mode, event, isZoomXDisabled)].onDragMove(event),
  onDragEnd: (event) => interactions[(0, import_get_active_mode.getActiveMode)(mode, event, isZoomXDisabled)].onDragEnd(event),
  onDragStartWithMiddleButton: () => interactions[(0, import_get_active_mode.getActiveMode)(mode)].onDragStartWithMiddleButton(),
  onDragMoveWithMiddleButton: (event) => interactions[(0, import_get_active_mode.getActiveMode)(mode, event, isZoomXDisabled)].onDragMoveWithMiddleButton(event),
  onDragEndWithMiddleButton: (event) => interactions[(0, import_get_active_mode.getActiveMode)(mode, event, isZoomXDisabled)].onDragEndWithMiddleButton(event),
  onTap: (event) => interactions[(0, import_get_active_mode.getActiveMode)(mode, event)].onMouseClickOrTap(event),
  onDoubleTap: () => interactions[(0, import_get_active_mode.getActiveMode)(mode)].onDoubleClickOrTap("pointer"),
  onMouseClick: (event) => interactions[(0, import_get_active_mode.getActiveMode)(mode, event, isZoomDisabled)].onMouseClickOrTap(
    event
  ),
  onMouseDoubleClick: () => interactions[(0, import_get_active_mode.getActiveMode)(mode)].onDoubleClickOrTap("pointer"),
  onMouseHover: (event) => interactions[(0, import_get_active_mode.getActiveMode)(mode, event, isZoomXDisabled, isPanDisabled)].onMouseHover(event)
});
