// @dynatrace/strato-components/charts/core/components/interactions-overlay/utils/get-pan-interactions.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var get_pan_interactions_exports = {};
__export(get_pan_interactions_exports, {
  getPanInteractions: () => getPanInteractions
});
module.exports = __toCommonJS(get_pan_interactions_exports);
var import_dragInteractions = require("../interactions/pan/dragInteractions.js");
var import_keyboardInteractions = require("../interactions/pan/keyboardInteractions.js");
var import_mouseInteractions = require("../interactions/pan/mouseInteractions.js");
var import_keyboardInteractions2 = require("../interactions/zoom/keyboardInteractions.js");
var import_reset = require("./reset.js");
const getPanInteractions = ({
  panDeltaRef,
  onPanStart,
  selectionArea,
  eventCancelled,
  onSelectEnd,
  lastMousePositionRef,
  pinned,
  onHover,
  onTogglePin,
  setCursorMode,
  height,
  onReset,
  width,
  onZoomChange,
  onModeChange,
  onPanEnd,
  onPanChange,
  cursorMode
}) => {
  return {
    onMouseHover: (event) => (0, import_mouseInteractions.panMouseHover)({
      event,
      pinned,
      lastMousePositionRef,
      cursorMode,
      setCursorMode,
      onHover
    }),
    onMouseClickOrTap: (event) => (0, import_mouseInteractions.panMouseClickOrTap)({ event, lastMousePositionRef }),
    onDragStart: (event) => (0, import_dragInteractions.panDragStart)({ panDeltaRef, onPanStart }),
    onDragMove: (event) => (0, import_dragInteractions.panDragMove)({ event, height, panDeltaRef, onPanChange }),
    onDragEnd: (event) => (0, import_dragInteractions.panDragEnd)({ panDeltaRef, onPanEnd }),
    onDragStartWithMiddleButton: () => (0, import_dragInteractions.panDragStartWithMiddleButton)({ panDeltaRef, onPanStart }),
    onDragMoveWithMiddleButton: (event) => (0, import_dragInteractions.panDragMoveWithMiddleButton)({ event, panDeltaRef, onPanChange }),
    onDragEndWithMiddleButton: (event) => (0, import_dragInteractions.panDragEndWithMiddleButton)({
      event,
      panDeltaRef,
      lastMousePositionRef,
      onPanEnd
    }),
    onDoubleClickOrTap: (triggeredBy) => {
      onReset?.({
        triggeredBy
      });
      (0, import_reset.reset)(lastMousePositionRef, panDeltaRef);
    },
    onWheel: (event) => (0, import_mouseInteractions.panMouseWheel)({ event, onPanChange }),
    onKeyEscape: () => (0, import_keyboardInteractions.panKeyboardEscape)({
      selectionArea,
      eventCancelled,
      onSelectEnd,
      lastMousePositionRef,
      pinned,
      onHover,
      onTogglePin,
      onModeChange,
      onPanEnd
    }),
    onKeyEnter: () => {
      return;
    },
    onKeyRight: (shiftKey) => (0, import_keyboardInteractions.panKeyboardRight)({ shiftKey, width, onPanChange }),
    onKeyLeft: (shiftKey) => (0, import_keyboardInteractions.panKeyboardLeft)({ shiftKey, width, onPanChange }),
    onKeyUp: (isTemporaryZoomMode) => {
      if (isTemporaryZoomMode) {
        (0, import_keyboardInteractions2.zoomKeyboardUp)({ width, onZoomChange });
      }
    },
    onKeyDown: (isTemporaryZoomMode) => {
      if (isTemporaryZoomMode) {
        (0, import_keyboardInteractions2.zoomKeyboardDown)({ width, onZoomChange });
      }
    }
  };
};
