// @dynatrace/strato-components/forms/search-input/SearchInput.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SearchInput_exports = {};
__export(SearchInput_exports, {
  SearchInput: () => SearchInput
});
module.exports = __toCommonJS(SearchInput_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_strato_icons = require("@dynatrace/strato-icons");
var styles = __toESM(require("./SearchInput.sty.js"));
var import_Stepper = require("./Stepper.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useControllableState = require("../../core/hooks/useControllableState.js");
var import_useSlots = require("../../core/hooks/useSlots.js");
var import_merge_props = require("../../core/utils/merge-props.js");
var import_BaseInput = require("../_base-input/BaseInput.js");
var import_BaseInputButton = require("../_base-input/BaseInputButton.js");
var import_BaseInputContent = require("../_base-input/BaseInputContent.js");
var import_ClearButton = require("../_base-input/ClearButton.js");
const COMPONENT_NAME = "SearchInput";
const SearchInputRoot = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    children,
    onChange,
    value,
    defaultValue,
    placeholder,
    required,
    readOnly = false,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "aria-disabled": ariaDisabledProp,
    className: consumerClassName,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const [internalValue, setInternalValue] = (0, import_useControllableState.useControllableState)({
    defaultValue,
    value,
    onChange
  });
  const handleOnChange = (e) => {
    setInternalValue(e.target.value);
    if (slots.stepper && e.target.value === "") {
      const { onCurrentChange } = slots.stepper.props;
      onCurrentChange(0);
    }
  };
  const handleKeyDown = (e) => {
    if (!slots.stepper) {
      return;
    }
    if (e.key !== "Enter") {
      return;
    }
    e.preventDefault();
    const { current, total, onCurrentChange } = slots.stepper.props;
    if (e.shiftKey) {
      onCurrentChange(current > 0 ? current - 1 : total - 1);
    } else {
      onCurrentChange(current < total - 1 ? current + 1 : 0);
    }
  };
  function onClear() {
    setInternalValue("");
  }
  const [slots] = (0, import_useSlots.useSlots)(children, {
    stepper: import_Stepper.Stepper
  });
  const isEmpty = internalValue === "" || internalValue === void 0;
  const hasStepper = !!slots.stepper;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_BaseInput.BaseInput,
    {
      ref: forwardedRef,
      ...(0, import_merge_props.mergeProps)(restProps, behavioralTrackingProps),
      ...hasStepper ? { onKeyDown: handleKeyDown } : {},
      className: (0, import_clsx.default)(styles.searchInput({ hasStepper }), consumerClassName),
      type: "search",
      value: internalValue,
      required,
      readOnly,
      onChange: handleOnChange,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "aria-disabled": ariaDisabledProp,
      placeholder,
      "data-dt-component": COMPONENT_NAME,
      size: 1,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_BaseInput.BaseInput.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.MagnifyingGlassIcon, {}) }),
        !isEmpty && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_BaseInput.BaseInput.ReservedSuffix, { children: [
          slots.stepper,
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ClearButton.ClearButton, { onClear })
        ] }),
        children
      ]
    }
  );
});
const SearchInput = Object.assign(SearchInputRoot, {
  Suffix: import_BaseInputContent.BaseInputSuffix,
  Button: import_BaseInputButton.BaseInputButton,
  Stepper: import_Stepper.Stepper
});
SearchInput.displayName = COMPONENT_NAME;
