// @dynatrace/strato-components/forms/search-input/Stepper.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Stepper_exports = {};
__export(Stepper_exports, {
  Stepper: () => Stepper
});
module.exports = __toCommonJS(Stepper_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var styles = __toESM(require("./Stepper.sty.js"));
var import_Button = require("../../buttons/button/Button.js");
var import_logging = require("../../core/utils/logging.js");
var import_Divider = require("../../layouts/divider/Divider.js");
var import_Flex = require("../../layouts/flex/Flex.js");
var baseInputStyles = __toESM(require("../_base-input/BaseInput.sty.js"));
var import_BaseInputContext = require("../_base-input/contexts/BaseInputContext.js");
const Stepper = ({
  current,
  onCurrentChange,
  total
}) => {
  const intl = (0, import_react_intl.useIntl)();
  const { hasError, disabled, ariaDisabled } = (0, import_react.useContext)(import_BaseInputContext.BaseInputContext);
  const disabledOrAriaDisabled = disabled || ariaDisabled || false;
  if (current > total) {
    (0, import_logging.devLog)(
      `The current match index (${current}) is greater than the total number of matches (${total}). Please ensure that the current index is zero-based and less than the total.`
    );
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { gap: 4, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      "div",
      {
        className: styles.matchPositionIndicators,
        role: "status",
        "aria-live": "polite",
        "aria-label": intl.formatMessage(
          {
            id: "tw9ulrSgegYXHq69",
            defaultMessage: "Match {current} of {total}",
            description: "Label for the current match position"
          },
          { current: current + 1, total }
        ),
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { children: current + 1 }),
          " / ",
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { children: total })
        ]
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Divider.Divider, { orientation: "vertical", className: baseInputStyles.divider }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { className: styles.stepperButtons, role: "group", children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_Button.Button,
        {
          onClick: (e) => {
            e.stopPropagation();
            onCurrentChange(current === 0 ? total - 1 : current - 1);
          },
          className: styles.stepperButton,
          color: hasError ? "critical" : "neutral",
          disabled: disabledOrAriaDisabled || total === 0,
          "aria-label": intl.formatMessage({
            id: "GLK6cN59tiRO9GDp",
            defaultMessage: "Previous match",
            description: "Label for the previous match button"
          }),
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ChevronUpIcon, {}) })
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_Button.Button,
        {
          onClick: (e) => {
            e.stopPropagation();
            onCurrentChange(current === total - 1 ? 0 : current + 1);
          },
          className: styles.stepperButton,
          color: hasError ? "critical" : "neutral",
          disabled: disabledOrAriaDisabled || total === 0,
          "aria-label": intl.formatMessage({
            id: "JLMJNMOPi2dCAbD1",
            defaultMessage: "Next match",
            description: "Label for the next match button"
          }),
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ChevronDownIcon, {}) })
        }
      )
    ] })
  ] });
};
Stepper.displayName = "SearchInput.Stepper";
