// @dynatrace/strato-components/esm/charts/core/components/text-ellipsis/truncate-string.js
import { ELLIPSIS_CHARACTER as _ELLIPSIS_CHARACTER } from "../../../../typography/utils.js";
const TEXT_HEIGHT = 12;
const calculateWidth = (text, fontWidthMap) => {
  let totalWidth = 0;
  for (const char of text) {
    totalWidth += fontWidthMap[char] ?? fontWidthMap["@"];
  }
  return totalWidth;
};
function accumulateUntilCapacity(availableWidth, fontWidthMap, generator) {
  let keptText = "";
  let keptWidth = 0;
  const capacity = availableWidth - calculateWidth(_ELLIPSIS_CHARACTER, fontWidthMap);
  for (const ch of generator) {
    const chWidth = calculateWidth(ch, fontWidthMap);
    if (keptWidth + chWidth > capacity) {
      return keptText;
    }
    keptText += ch;
    keptWidth += chWidth;
  }
  return keptText;
}
const EPSILON_PX = 0.5;
const MIN_AVAILABLE_WIDTH = 19;
const buildTruncatedOutput = {
  start: (kept) => {
    let reversed = "";
    for (const ch of kept) {
      reversed = ch + reversed;
    }
    return `${_ELLIPSIS_CHARACTER}${reversed}`;
  },
  end: (kept) => `${kept}${_ELLIPSIS_CHARACTER}`,
  middle: (kept, priority) => {
    let head = "";
    let tail = "";
    if (priority === "start") {
      for (let i = 0; i < kept.length; i++) {
        if (i % 2 === 0) {
          head += kept[i];
        } else {
          tail = kept[i] + tail;
        }
      }
    } else {
      for (let i = 0; i < kept.length; i++) {
        if (i % 2 === 0) {
          tail = kept[i] + tail;
        } else {
          head += kept[i];
        }
      }
    }
    return `${head}${_ELLIPSIS_CHARACTER}${tail}`;
  }
};
const keepSequenceGenerators = {
  start: function* fromTail(text) {
    for (let i = text.length - 1; i >= 0; i--) {
      yield text[i];
    }
  },
  end: function* fromHead(text) {
    for (const ch of text) {
      yield ch;
    }
  },
  middle: function* fromBothEnds(text, priority) {
    let left = 0;
    let right = text.length - 1;
    while (left < right) {
      if (priority === "start") {
        yield text[left++];
        yield text[right--];
      } else {
        yield text[right--];
        yield text[left++];
      }
    }
    if (left === right) {
      yield text[left];
    }
  }
};
const truncateString = (text, availableWidth, fontWidthMap, truncationMode = "middle", priority = "end") => {
  const fullTextWidth = calculateWidth(text, fontWidthMap);
  const roundedAvailableWidth = Math.round(availableWidth) + EPSILON_PX;
  if (fullTextWidth <= roundedAvailableWidth) {
    return text;
  }
  if (MIN_AVAILABLE_WIDTH >= roundedAvailableWidth) {
    return _ELLIPSIS_CHARACTER;
  }
  const keptChars = accumulateUntilCapacity(
    availableWidth,
    fontWidthMap,
    truncationMode === "middle" ? keepSequenceGenerators.middle(text, priority) : keepSequenceGenerators[truncationMode](text)
  );
  return truncationMode === "middle" ? buildTruncatedOutput.middle(keptChars, priority) : buildTruncatedOutput[truncationMode](keptChars);
};
export {
  TEXT_HEIGHT,
  calculateWidth,
  truncateString
};
