// @dynatrace/strato-components/esm/overlays/drawer/Drawer.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import { AnimatePresence, motion } from "framer-motion";
import {
  forwardRef,
  useRef
} from "react";
import AnimationsJS from "@dynatrace/strato-design-tokens/animations-js";
import Colors from "@dynatrace/strato-design-tokens/colors";
import {
  DrawerBackdropCSS,
  DrawerCSS,
  DrawerContentCSS,
  DrawerPlacementCSS,
  DrawerWidthStyles
} from "./Drawer.sty.js";
import { FocusScope } from "../../core/components/focus-scope/FocusScope.js";
import { OverlayContainer } from "../../core/components/OverlayContainer.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useMergeRefs } from "../../core/hooks/useMergeRefs.js";
import { useOverlayStacking } from "../../core/hooks/useOverlayStacking.js";
import { OverlayStackingProvider } from "../../core/providers/OverlayStackingProvider.js";
import {
  fadeInEasing,
  fadeOutEasing,
  slideInEasing,
  slideOutEasing
} from "../../core/utils/framer-motion-easings.js";
import { useOverlay } from "../hooks/useOverlay.js";
const COMPONENT_NAME = "Drawer";
const slideAnimation = ($placement) => ({
  visible: {
    x: "0",
    transition: {
      duration: AnimationsJS.SlideIn.Duration.Slow,
      ease: slideInEasing
    },
    // Workaround to fix remaining style (transform: translateX(0)) on Drawer
    // as this breaks any child element with position: fixed
    // https://github.com/framer/motion/issues/823
    transitionEnd: {
      x: 0
    }
  },
  hidden: {
    x: $placement === "left" ? "-100%" : "100%",
    transition: {
      duration: AnimationsJS.SlideOut.Duration.Slow,
      ease: slideOutEasing
    }
  }
});
const fade = {
  fadeIn: {
    // For the opacity we need to use the default value here. Framer motion would complain
    // about too complex values, when transitioning from two custom property values within
    // an rgba value apparently.
    opacity: 1,
    transition: {
      duration: AnimationsJS.SlideIn.Duration.Slow,
      ease: fadeInEasing
    }
  },
  fadeOut: {
    opacity: 0,
    transition: {
      duration: AnimationsJS.SlideOut.Duration.Slow,
      ease: fadeOutEasing
    }
  }
};
const Drawer = /* @__PURE__ */ forwardRef(
  (props, forwardedRef) => {
    const {
      placement = "right",
      width = "medium",
      isDismissed,
      onDismiss,
      children,
      className: consumerClassName,
      style: consumerStyle,
      modal = true,
      "data-testid": dataTestId,
      disableAnimations = false,
      ...remainingProps
    } = props;
    const ref = useRef(null);
    const { overlayProps } = useOverlay(
      {
        onClose: (reason) => {
          onDismiss(true, reason);
        },
        isOpen: !isDismissed,
        isDismissible: false,
        isKeyboardDismissDisabled: false
      },
      ref
    );
    const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
    const mergedRefs = useMergeRefs([ref, forwardedRef]);
    const overlayContainerProps = useOverlayStacking(0, "overlay");
    return /* @__PURE__ */ jsx(AnimatePresence, { children: !isDismissed && /* @__PURE__ */ jsx(
      OverlayContainer,
      {
        "data-testid": "drawer-container",
        ...overlayContainerProps,
        style: {
          ...overlayContainerProps.style,
          pointerEvents: modal ? void 0 : "none"
        },
        children: /* @__PURE__ */ jsx(
          DrawerBackdrop,
          {
            modal,
            overlayProps,
            onDismiss,
            disableAnimations,
            children: /* @__PURE__ */ jsx(
              motion.div,
              {
                "data-testid": dataTestId,
                ref: mergedRefs,
                variants: disableAnimations ? {} : slideAnimation(placement),
                initial: "hidden",
                animate: "visible",
                exit: "hidden",
                onAnimationComplete: () => window.dispatchEvent(new Event("animationDone")),
                tabIndex: -1,
                className: clsx(
                  consumerClassName,
                  DrawerCSS,
                  DrawerPlacementCSS[placement],
                  {
                    [DrawerWidthStyles.narrow]: width === "narrow",
                    [DrawerWidthStyles.medium]: width === "medium",
                    [DrawerWidthStyles.wide]: width === "wide",
                    [DrawerWidthStyles.full]: width === "full"
                  }
                ),
                style: {
                  width: (
                    // Avoid binding the predefined named widths to the style, but add any other values
                    width !== "narrow" && width !== "medium" && width !== "wide" && width !== "full" ? width : void 0
                  ),
                  ...consumerStyle
                },
                ...restProps,
                ...behavioralTrackingProps,
                "data-dt-component": COMPONENT_NAME,
                children: /* @__PURE__ */ jsx(FocusScope, { autoFocus: true, contain: modal, children: /* @__PURE__ */ jsx(
                  motion.div,
                  {
                    variants: disableAnimations ? {} : fade,
                    initial: "fadeOut",
                    animate: "fadeIn",
                    exit: "fadeOut",
                    className: DrawerContentCSS,
                    children: /* @__PURE__ */ jsx(OverlayStackingProvider, { children })
                  }
                ) })
              }
            )
          }
        )
      }
    ) });
  }
);
const DrawerBackdrop = (props) => {
  const { children, modal, disableAnimations, onDismiss, overlayProps } = props;
  return /* @__PURE__ */ jsx(
    motion.div,
    {
      variants: disableAnimations ? {} : fade,
      initial: "fadeOut",
      animate: "fadeIn",
      exit: "fadeOut",
      "data-testid": "drawer-backdrop",
      className: DrawerBackdropCSS,
      style: {
        pointerEvents: modal ? void 0 : "none",
        backgroundColor: modal ? Colors.Background.Surface.Backdrop : void 0
      },
      onPointerDown: (event) => {
        if (modal === true && event.target === event.currentTarget) {
          onDismiss(true, "backdrop");
        }
      },
      ...overlayProps,
      children
    }
  );
};
export {
  Drawer
};
