// @dynatrace/strato-components/esm/tables/SimpleTable/hooks/useColumns.js
import { useMemo } from "react";
function isSimpleTableGroupColumnDef(columnDef) {
  return columnDef !== null && typeof columnDef === "object" && "columns" in columnDef && Array.isArray(columnDef.columns);
}
function isSimpleTableDisplayColumnDef(columnDef) {
  return columnDef !== null && typeof columnDef === "object" && "accessor" in columnDef;
}
function splitAccessor(column) {
  if (typeof column.accessor === "string") {
    return {
      accessorKey: column.accessor
    };
  }
  if (typeof column.accessor === "function") {
    return {
      accessorFn: column.accessor
    };
  }
  return {};
}
function createHeaderRenderer(header) {
  if (typeof header === "function") {
    return () => header();
  }
  return header;
}
function createDisplayColumnDef(columnDef) {
  const composedDef = {
    id: columnDef.id,
    header: createHeaderRenderer(columnDef.header),
    meta: {
      width: columnDef.width,
      minWidth: columnDef.minWidth,
      maxWidth: columnDef.maxWidth,
      alignment: columnDef.alignment ?? "left"
    },
    ...splitAccessor(columnDef)
  };
  if (columnDef.cell) {
    const customCell = columnDef.cell;
    if (typeof customCell === "string") {
      composedDef.cell = customCell;
      return composedDef;
    }
    if (typeof customCell === "function") {
      return {
        ...composedDef,
        cell: createCellRenderer(customCell)
      };
    }
    return composedDef;
  }
  return composedDef;
}
function createCellRenderer(consumerCustomCell) {
  return ({ getValue, row, column }) => {
    return consumerCustomCell({
      value: getValue(),
      rowIndex: row.index,
      rowData: row.original,
      rowId: row.id
    });
  };
}
function useColumns(propColumns) {
  return useMemo(() => {
    const createdColumns = [];
    for (const firstLevelColumn of propColumns) {
      if (isSimpleTableGroupColumnDef(firstLevelColumn)) {
        const groupColumn = {
          id: firstLevelColumn.id,
          header: createHeaderRenderer(firstLevelColumn.header),
          columns: firstLevelColumn.columns?.map(
            (childColumn) => createDisplayColumnDef(childColumn)
          ),
          meta: {
            width: firstLevelColumn.width,
            alignment: firstLevelColumn.alignment ?? "left"
          }
        };
        createdColumns.push(groupColumn);
      } else if (isSimpleTableDisplayColumnDef(firstLevelColumn)) {
        createdColumns.push(createDisplayColumnDef(firstLevelColumn));
      }
    }
    return createdColumns;
  }, [propColumns]);
}
export {
  isSimpleTableGroupColumnDef,
  useColumns
};
