// @dynatrace-sdk/app-environment/cjs/index.js
/**
 * @license
 * Copyright 2023-2025 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/platform/app-environment/src/index.ts
var src_exports = {};
__export(src_exports, {
  getAppId: () => getAppId,
  getAppName: () => getAppName,
  getAppVersion: () => getAppVersion,
  getCurrentUserDetails: () => getCurrentUserDetails,
  getEnvironmentId: () => getEnvironmentId,
  getEnvironmentUrl: () => getEnvironmentUrl
});
module.exports = __toCommonJS(src_exports);

// packages/platform/shared/utils/get-global-with-dt-runtime.ts
var getGlobalWithDtRuntime = () => typeof globalThis !== "undefined" ? globalThis : window;

// packages/platform/app-environment/src/lib/get-app-id.ts
function getAppId() {
  const runtimeFunc = getGlobalWithDtRuntime().dtRuntime?.appEnvironment?.getAppId;
  if (typeof runtimeFunc === "function") {
    return runtimeFunc();
  } else {
    console.warn("Missing getAppId function from web runtime.");
    return "dt.missing.app.id";
  }
}

// packages/platform/app-environment/src/lib/get-app-name.ts
function getAppName() {
  const runtimeFunc = getGlobalWithDtRuntime().dtRuntime?.appEnvironment?.getAppName;
  if (typeof runtimeFunc === "function") {
    return runtimeFunc();
  } else {
    console.warn("Missing getAppName function from web runtime.");
    return "dt.missing.app.name";
  }
}

// packages/platform/app-environment/src/lib/get-app-version.ts
function getAppVersion() {
  const runtimeFunc = getGlobalWithDtRuntime().dtRuntime?.appEnvironment?.getAppVersion;
  if (typeof runtimeFunc === "function") {
    return runtimeFunc();
  } else {
    console.warn("Missing getAppVersion function from web runtime.");
    return "dt.missing.app.version";
  }
}

// packages/platform/app-environment/src/lib/get-environment-id.ts
function getEnvironmentId() {
  const runtimeFunc = getGlobalWithDtRuntime().dtRuntime?.appEnvironment?.getEnvironmentId;
  if (typeof runtimeFunc === "function") {
    return runtimeFunc();
  } else {
    console.warn("Missing getEnvironmentId function from web runtime.");
    return "dt.missing.environment.id";
  }
}

// packages/platform/app-environment/src/lib/get-environment-url.ts
function getEnvironmentUrl() {
  const runtimeFunc = getGlobalWithDtRuntime().dtRuntime?.appEnvironment?.getEnvironmentUrl;
  if (typeof runtimeFunc === "function") {
    return runtimeFunc();
  } else {
    console.warn("Missing getEnvironmentUrl function from web runtime.");
    return "https://dynatrace.com/";
  }
}

// packages/platform/app-environment/src/lib/get-current-user-details.ts
function getCurrentUserDetails() {
  const runtimeFunc = getGlobalWithDtRuntime().dtRuntime?.appEnvironment?.getCurrentUserDetails;
  if (typeof runtimeFunc === "function") {
    return runtimeFunc();
  } else {
    console.warn("Missing getCurrentUserDetails function from web runtime.");
    return {
      id: "dt.missing.user.id",
      name: "dt.missing.user.name",
      email: "dt.missing.user.email"
    };
  }
}
