// @dynatrace/strato-components/esm/tables/DataTable/components/DataTableHeaderCell.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { createElement } from "react";
import { flexRender } from "@tanstack/react-table";
import { assignInlineVars } from "@vanilla-extract/dynamic";
import clsx from "clsx";
import {
  memo,
  useCallback,
  useRef
} from "react";
import { createPortal } from "react-dom";
import { useIntl } from "react-intl";
import * as alignmentStyles from "./Alignment.sty.js";
import { DataTablePinningShadowHeaderCells } from "./ColumnPinning/DataTablePinningShadowHeaderCells.js";
import * as headerCellStyles from "./DataTableHeaderCell.sty.js";
import { useMergeRefs } from "../../../core/hooks/useMergeRefs.js";
import { Tooltip } from "../../../overlays/tooltip/Tooltip.js";
import { TextEllipsis } from "../../../typography/text-ellipsis/TextEllipsis.js";
import { getDataTableBorderWidth } from "../../utils/data-table-border-width.js";
import { getColumnLabel } from "../../utils/get-column-label.js";
import {
  DATA_TABLE_DEFAULT_HORIZONTAL_ALIGNMENT,
  DATA_TABLE_SPACER_COLUMN_ID
} from "../constants.js";
import {
  DataTableColumnContextProvider,
  DataTableHeaderContextProvider,
  useDataTable
} from "../contexts/DataTableContext.js";
import * as tableStyles from "../DataTable.sty.js";
import * as columnOrderStyles from "../features/ColumnOrder/ColumnOrder.sty.js";
import { SortingArrowIndicators } from "../features/Sorting/SortingArrowIndicators.js";
import { SubRowsHeaderExpander } from "../features/SubRows/SubRowExpanders.js";
import { SubRowsHeaderWidthCorrection } from "../features/SubRows/SubRows.js";
import { InlineUserActionsMenu } from "../features/UserActions/InlineUserActionsMenu.js";
import { useHeaderCellInteractions } from "../hooks/useHeaderCellInteractions.js";
import { useHeaderCellDragAndDrop } from "../hooks/useHeaderDragAndDrop.js";
import { dataTableMessages } from "../messages.js";
const HeaderTextEllipsis = memo(
  ({
    children,
    onTextOverflow
  }) => {
    return /* @__PURE__ */ jsx(
      TextEllipsis,
      {
        onTextOverflow,
        className: headerCellStyles.headerCellEllipsis,
        children
      }
    );
  }
);
function getHeaderAriaLabel({
  intl,
  columnLabel,
  nextSortingOrder,
  canSort,
  columnDragAndDropMode,
  isCustomHeader,
  role
}) {
  if (role === "presentation") {
    return void 0;
  }
  if (canSort && columnDragAndDropMode === "none") {
    return intl.formatMessage(dataTableMessages.sortingLabel, {
      column: columnLabel,
      direction: nextSortingOrder
    });
  }
  if (canSort && columnDragAndDropMode === "enabled") {
    return intl.formatMessage(dataTableMessages.sortingDragAndDropLabel, {
      column: columnLabel,
      direction: nextSortingOrder
    });
  }
  if (columnDragAndDropMode !== "none") {
    return intl.formatMessage(dataTableMessages.dragAndDropLabel, {
      column: columnLabel
    });
  }
  return isCustomHeader ? columnLabel : void 0;
}
function getHeaderAriaSortDescription({
  intl,
  role,
  isSorted,
  multiSortPriority,
  sortedColumnsCount
}) {
  if (role === "presentation" || !isSorted) {
    return void 0;
  }
  if (multiSortPriority !== void 0) {
    return intl.formatMessage(dataTableMessages.sortingStateWithPriorityLabel, {
      direction: isSorted,
      priority: multiSortPriority,
      total: sortedColumnsCount
    });
  }
  return intl.formatMessage(dataTableMessages.sortingStateLabel, {
    direction: isSorted
  });
}
function getInteractionDtName({
  columnDragAndDropMode,
  canSort
}) {
  if (columnDragAndDropMode === "only" || columnDragAndDropMode === "enabled" && !canSort) {
    return "ColumnHeaderOrder";
  }
  if (canSort && columnDragAndDropMode === "enabled") {
    return "ColumnHeaderOrderAndSort";
  }
  return "ColumnHeaderSort";
}
function getInteractionDtProperties(canSort, nextSortingOrder) {
  if (!canSort || nextSortingOrder === void 0) {
    return void 0;
  }
  switch (nextSortingOrder) {
    case "asc":
      return "state:ascending";
    case "desc":
      return "state:descending";
    case false:
      return "state:clearSort";
  }
}
function getLeftAdditionalElevation(isPinnedLeft, canResize, columnId, leftPinnedColumns) {
  if (!isPinnedLeft || !canResize) {
    return "0";
  }
  const idx = leftPinnedColumns.findIndex((c) => c.id === columnId);
  return `${leftPinnedColumns.length - 1 - idx}`;
}
function DataTableHeaderCell(props) {
  const {
    header,
    headers,
    headerGroupId,
    rowIndex,
    columnIndex: virtualColumnIndex,
    lastRowIndex
  } = props;
  const intl = useIntl();
  const table = useDataTable();
  const { rowDensity, verticalDividers } = table.getVisualVariantOptions();
  const columnDragAndDropMode = header.getColumnDragAndDropMode();
  const gridPosition = header.getGridPosition(headers);
  const column = header.column;
  const { id: columnId, parent, columns: childColumns, columnDef } = column;
  const columnIndex = rowIndex === 0 ? gridPosition.gridColumnStart - 1 : virtualColumnIndex;
  const canSort = column.getCanSort();
  const canColumnActions = column.getCanColumnActions();
  const textOverflowRef = useRef(false);
  const handleTextOverflow = useCallback((ellipsized) => {
    textOverflowRef.current = ellipsized;
  }, []);
  const { setNodeRef, isDragging, portalRef, dragAndDropProps } = useHeaderCellDragAndDrop({
    headerId: header.id,
    column,
    columnIndex,
    columnDragAndDropMode
  });
  const divRef = useRef(null);
  const mergedRef = useMergeRefs([setNodeRef, divRef]);
  const tableState = table.getState();
  const isActiveFloatingMenu = tableState.floatingMenuInfo?.rowId === null && tableState.floatingMenuInfo?.headerId === header.id;
  const {
    handleCellMouseEnter,
    handleCellMouseLeave,
    handleCellActionsClose,
    handleCellClick
  } = useHeaderCellInteractions({
    headerCell: divRef.current,
    canSort,
    canColumnActions,
    isActiveFloatingMenu,
    rowDensity,
    getToggleSortingHandler: column.getToggleSortingHandler
  });
  if (header.isPlaceholder) {
    return null;
  }
  const headerGroupsLength = table.getHeaderGroups().length;
  const activeSortEntries = tableState.sorting?.filter((sort) => {
    const col = table.getColumn(sort.id);
    return col?.getCanSort() ?? false;
  }) ?? [];
  const sortedColumnsCount = activeSortEntries.length;
  const enableExpanding = table.options.enableExpanding;
  const canResize = column.getCanResize();
  const isResizing = column.getIsResizing();
  const isSorted = column.getIsSorted();
  const columnSortIndex = activeSortEntries.findIndex(
    (sort) => sort.id === column.id
  );
  const multiSortPriority = sortedColumnsCount > 1 && columnSortIndex >= 0 ? columnSortIndex + 1 : void 0;
  const isColumnDragActive = tableState.draggedHeaderColumn !== void 0;
  const isRightAligned = columnDef.alignment === "right";
  const renderVerticalDividers = verticalDividers || table.getHeaderGroups().length > 1 || table.options.enableColumnResizing && (canResize || // when resizing is enabled we need to render the vertical dividers for the spacer column as well
  column.id === DATA_TABLE_SPACER_COLUMN_ID);
  const hasCustomHeaderRenderer = columnDef.meta?.hasCustomHeaderRenderer;
  const children = hasCustomHeaderRenderer ? /* @__PURE__ */ jsx(DataTableColumnContextProvider, { value: header.column, children: flexRender(columnDef.header, header.getContext()) }) : flexRender(columnDef.header, header.getContext());
  const isStringHeader = typeof children === "string";
  const role = columnDef.meta?.spacer === true ? "presentation" : "columnheader";
  const columnLabel = getColumnLabel(column);
  const nextSortingOrder = column.getNextSortingOrder();
  const ariaLabel = getHeaderAriaLabel({
    intl,
    columnLabel,
    nextSortingOrder,
    canSort,
    columnDragAndDropMode,
    isCustomHeader: !isStringHeader,
    role
  });
  const ariaSortDescription = getHeaderAriaSortDescription({
    intl,
    role,
    isSorted,
    multiSortPriority,
    sortedColumnsCount
  });
  const isSubRowIndicatorColumn = enableExpanding === true && table.getSubRowIndicatorColumnId() === column.id;
  const subRowsWidthCorrection = (
    // only render the subrow width correction in the subrow indicator column
    enableExpanding === true && isSubRowIndicatorColumn ? /* @__PURE__ */ jsx(
      SubRowsHeaderWidthCorrection,
      {
        subRowOpenDepth: table.getSubRowOpenDepth()
      }
    ) : null
  );
  const pinningPosition = header.getIsPinned();
  const isPinnedLeft = pinningPosition === "left";
  const isPinnedRight = pinningPosition === "right";
  const leftPinnedPosition = isPinnedLeft ? header.getLeftPinnedPosition() : void 0;
  const leftPinnedPositionPx = leftPinnedPosition === void 0 ? void 0 : `${leftPinnedPosition}px`;
  const leftAdditionalElevation = getLeftAdditionalElevation(
    isPinnedLeft,
    canResize,
    columnId,
    table.getLeftVisibleLeafColumns()
  );
  const rightPinnedPosition = isPinnedRight ? header.getRightPinnedPosition() : void 0;
  const rightPinnedPositionPx = rightPinnedPosition === void 0 ? void 0 : `${rightPinnedPosition}px`;
  const hasChildren = childColumns.length > 0;
  const isEdgeColumnInGroup = parent?.getCenterVisibleLeafColumns().at(-1) === column || parent?.getLeftVisibleLeafColumns().at(-1) === column || parent?.getRightVisibleLeafColumns().at(0) === column;
  const isLastVisibleLeftPinnedHeader = isPinnedLeft ? header.getIsLastVisibleLeftPinnedHeader() : false;
  const isFirstVisibleRightPinnedHeader = isPinnedRight ? header.getIsFirstVisibleRightPinnedHeader() : false;
  const isFirstVisibleUnpinnedHeader = pinningPosition === false ? header.getIsFirstVisibleUnpinnedHeader() : false;
  const isLastUnpinnedColumnWithoutSpacer = columnId === table.getCenterVisibleLeafColumns().at(-2)?.id;
  const edgeAdjustedGridPosition = {
    gridColumnStart: gridPosition.gridColumnStart,
    gridColumnEnd: gridPosition.gridColumnEnd,
    gridRowStart: isEdgeColumnInGroup ? 1 : gridPosition.gridRowStart,
    gridRowEnd: isEdgeColumnInGroup ? `span ${headerGroupsLength}` : gridPosition.gridRowEnd
  };
  const interactionDtName = getInteractionDtName({
    columnDragAndDropMode,
    canSort
  });
  const interactionDtProperties = getInteractionDtProperties(
    canSort,
    nextSortingOrder
  );
  const renderSubRowHeaderExpander = () => {
    if (!enableExpanding || !isSubRowIndicatorColumn || hasCustomHeaderRenderer) {
      return null;
    }
    return /* @__PURE__ */ jsx(SubRowsHeaderExpander, { isAllRowsExpanded: table.getIsAllRowsExpanded() });
  };
  let headerContent = /* @__PURE__ */ createElement(
    "div",
    {
      ...dragAndDropProps,
      key: `Header-${columnId}`,
      ref: mergedRef,
      role,
      "data-headerid": header?.id,
      "data-header-row": rowIndex,
      "data-column": columnIndex,
      "data-columnid": columnId,
      "data-columntype": columnDef.columnType,
      "data-dt-name": interactionDtName,
      "data-dt-properties": interactionDtProperties,
      className: clsx(
        headerCellStyles.headerCell({
          rowDensity: hasCustomHeaderRenderer ? "customHeaderRenderer" : rowDensity,
          pinnedColumn: pinningPosition,
          verticalDividers: renderVerticalDividers,
          isDragAndDropDisabled: header.getIsDragAndDropDisabled()
        }),
        alignmentStyles.horizontalAlignment[columnDef.alignment ?? DATA_TABLE_DEFAULT_HORIZONTAL_ALIGNMENT],
        alignmentStyles.verticalAlignment[table.getHeaderVerticalAlignment()],
        {
          [headerCellStyles.firstColumn]: columnIndex === 0,
          [headerCellStyles.firstHeaderRow]: rowIndex === 0,
          [headerCellStyles.lastLeftPinnedColumn]: isLastVisibleLeftPinnedHeader,
          [headerCellStyles.firstUnpinnedColumn]: isFirstVisibleUnpinnedHeader,
          [headerCellStyles.firstRightPinnedColumn]: isFirstVisibleRightPinnedHeader,
          [headerCellStyles.lastHeaderRow]: rowIndex === lastRowIndex,
          [headerCellStyles.parentHeader]: hasChildren,
          [headerCellStyles.childHeader]: !!parent,
          [headerCellStyles.sortableHeader]: canSort,
          [headerCellStyles.isSortedHeader]: isSorted,
          [headerCellStyles.isMultiSortedHeader]: multiSortPriority !== void 0,
          [tableStyles.prefixColumn]: columnDef.meta?.builtinPlacement === "prefix",
          [headerCellStyles.headerCellActiveInteraction]: isActiveFloatingMenu,
          [headerCellStyles.customHeaderDensity[rowDensity]]: hasCustomHeaderRenderer,
          [headerCellStyles.subRowIndicatorColumn]: isSubRowIndicatorColumn,
          // add grab cursor styling only if column drag and drop is enabled but sorting is not
          [columnOrderStyles.headerDragAndDrop.enabled]: !isDragging && columnDragAndDropMode !== "none" && !header.getCanSortByHeaderClick(),
          [columnOrderStyles.headerDragAndDrop.inactive]: tableState.draggedHeaderColumn !== void 0
        }
      ),
      "aria-sort": header.getAriaSortAttribute(),
      "aria-description": ariaSortDescription,
      "data-sort-desc-first": columnDef.sortDescFirst ?? false,
      "aria-label": ariaLabel,
      style: {
        ...assignInlineVars({
          [headerCellStyles.leftPositionVar]: leftPinnedPositionPx,
          [headerCellStyles.rightPositionVar]: rightPinnedPositionPx,
          [headerCellStyles.leftAdditionalElevationVar]: leftAdditionalElevation
        }),
        ...gridPosition
      },
      onMouseEnter: handleCellMouseEnter,
      onMouseLeave: handleCellMouseLeave,
      onClick: handleCellClick
    },
    renderSubRowHeaderExpander(),
    isRightAligned && subRowsWidthCorrection,
    isStringHeader ? /* @__PURE__ */ jsx(HeaderTextEllipsis, { onTextOverflow: handleTextOverflow, children }) : children,
    canSort && /* @__PURE__ */ jsx(
      SortingArrowIndicators,
      {
        isSorted,
        nextSortingOrder,
        multiSortPriority,
        rowDensity
      }
    ),
    canColumnActions && !isColumnDragActive && /* @__PURE__ */ jsx(DataTableHeaderContextProvider, { value: header, children: /* @__PURE__ */ jsx(
      InlineUserActionsMenu,
      {
        headerGroupOrRowId: headerGroupId,
        columnId,
        type: "header",
        "aria-label": intl.formatMessage(dataTableMessages.columnActionLabel),
        "data-header-row": rowIndex,
        "data-column": columnIndex,
        onClose: handleCellActionsClose
      }
    ) }),
    !isRightAligned && subRowsWidthCorrection,
    canResize && !isColumnDragActive && /* @__PURE__ */ jsx(
      "button",
      {
        type: "button",
        "aria-label": intl.formatMessage(dataTableMessages.resizeColumnLabel, {
          column: columnLabel
        }),
        "data-header-row": rowIndex,
        "data-column": columnIndex,
        "data-trigger": "resize",
        className: clsx(
          headerCellStyles.columnSizingHandle({
            active: isResizing,
            last: isLastUnpinnedColumnWithoutSpacer,
            pinnedColumn: pinningPosition
          }),
          {
            [headerCellStyles.firstRightPinnedColumn]: isFirstVisibleRightPinnedHeader
          }
        ),
        ...isEdgeColumnInGroup && {
          style: {
            // if this is the last column of a header group, the resize handle should stretch the full header height
            height: table.getTotalHeaderHeight() - getDataTableBorderWidth()
          }
        },
        onMouseDown: header.handleResize,
        onTouchStart: header.handleResize,
        onKeyUp: header.handleKeyboardResize,
        onDoubleClick: header.handleResizeExpand,
        "data-dt-name": "ColumnHeaderResize"
      },
      `Header-${columnId}-resize-handler`
    )
  );
  if (isDragging) {
    headerContent = createPortal(
      headerContent,
      portalRef.current || document.body
    );
  }
  if (isStringHeader) {
    headerContent = /* @__PURE__ */ jsx(Tooltip, { disabled: !textOverflowRef.current, text: children, children: headerContent });
  }
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    headerContent,
    /* @__PURE__ */ jsx(
      DataTablePinningShadowHeaderCells,
      {
        isLeftShadow: isLastVisibleLeftPinnedHeader && !hasChildren,
        isRightShadow: isFirstVisibleRightPinnedHeader && !hasChildren,
        leftPosition: leftPinnedPosition,
        rightPosition: rightPinnedPosition,
        gridPosition: edgeAdjustedGridPosition
      }
    )
  ] });
}
DataTableHeaderCell.displayName = "DataTableHeaderCell";
export {
  DataTableHeaderCell,
  getHeaderAriaLabel
};
