// @dynatrace/strato-components/esm/tables/DataTable/components/DataTableDefaultCell.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import { forwardRef } from "react";
import { DATA_TABLE_DEFAULT_HORIZONTAL_ALIGNMENT } from "../constants.js";
import * as alignmentStyles from "./Alignment.sty.js";
import * as cellStyles from "./DataTableCell.sty.js";
import * as defaultCellStyles from "./DataTableDefaultCell.sty.js";
import {
  useDataTable,
  useDataTableColumn,
  useDataTableIsHeaderContextProvider
} from "../contexts/DataTableContext.js";
import {
  SubRowsCellExpander,
  SubRowsHeaderExpander
} from "../features/SubRows/SubRowExpanders.js";
const DataTableDefaultCell = forwardRef((props, forwardedRef) => {
  const {
    children,
    style: consumerStyle,
    className: consumerClassName,
    "data-testid": dataTestId,
    ...remainingProps
  } = props;
  const table = useDataTable();
  const column = useDataTableColumn();
  const isHeader = useDataTableIsHeaderContextProvider();
  return /* @__PURE__ */ jsxs(
    "div",
    {
      ref: forwardedRef,
      className: clsx(
        defaultCellStyles.defaultCell({
          rowDensity: table.getVisualVariantOptions().rowDensity
        }),
        alignmentStyles.horizontalAlignment[column.columnDef.alignment ?? DATA_TABLE_DEFAULT_HORIZONTAL_ALIGNMENT],
        consumerClassName
      ),
      style: consumerStyle,
      "data-testid": dataTestId,
      ...remainingProps,
      children: [
        table.options.enableExpanding === true && table.getSubRowIndicatorColumnId() === column.columnDef.id && (isHeader ? /* @__PURE__ */ jsx(
          SubRowsHeaderExpander,
          {
            isAllRowsExpanded: table.getIsAllRowsExpanded()
          }
        ) : /* @__PURE__ */ jsx(SubRowsCellExpander, { columnType: column.columnDef.columnType })),
        !column.columnDef.meta?.builtin ? /* @__PURE__ */ jsx(
          "div",
          {
            className: clsx(
              cellStyles.innerCell({
                lineWrap: !isHeader && column.getIsLineWrapped()
              })
            ),
            children
          }
        ) : children
      ]
    }
  );
});
DataTableDefaultCell.displayName = "DataTable.DefaultCell";
export {
  DataTableDefaultCell
};
