// @dynatrace/strato-components/esm/tables/DataTable/components/DataTableBody.js
import { Fragment, jsx } from "react/jsx-runtime";
import clsx from "clsx";
import { useCallback, useRef } from "react";
import { DataTableRow } from "./DataTableRow.js";
import { useDataTable } from "../contexts/DataTableContext.js";
import { RowDetailsElement } from "../features/RowDetails/RowDetailsElement.js";
import { RowOrderingDragAndDropContext } from "../features/RowOrder/RowOrder.js";
import * as styles from "./DataTableBody.sty.js";
import {
  DATA_TABLE_DETAILS_ROW_IDENTIFIER,
  DATA_TABLE_DETAILS_ROW_PARITY,
  DATA_TABLE_ROW_EXPANDED_ID_SUFFIX
} from "../constants.js";
function DataTableBody(props) {
  const { rows, rowVirtualizer, columnVirtualizer } = props;
  const table = useDataTable();
  const queuedMeasurementRowIdRef = useRef(/* @__PURE__ */ new Set());
  const rowMeasure = useCallback(
    (node) => {
      if (!node || !(node instanceof HTMLElement)) {
        return;
      }
      const rowId = node.getAttribute("data-rowid");
      if (rowId === null || queuedMeasurementRowIdRef.current.has(rowId)) {
        return;
      }
      queuedMeasurementRowIdRef.current.add(rowId);
      requestAnimationFrame(() => {
        if (node.isConnected) {
          rowVirtualizer.measureElement(node);
        }
        queuedMeasurementRowIdRef.current.delete(rowId);
      });
    },
    [rowVirtualizer]
  );
  const columnMeasure = useCallback(
    (node) => {
      if (!node || !(node instanceof HTMLElement)) {
        return;
      }
      columnVirtualizer.measureElement(node);
    },
    [columnVirtualizer]
  );
  const firstRenderedRowIndex = rows.at(0)?.index;
  const rowSeparation = table.getVisualVariantOptions().rowSeparation;
  const virtualColumns = columnVirtualizer.getVirtualItems();
  const draggingRowId = table.getState().draggingRowInfo?.id;
  const draggingRowIdExpanded = `${draggingRowId}${DATA_TABLE_ROW_EXPANDED_ID_SUFFIX}`;
  const rowElements = rowVirtualizer.getVirtualItems().map((virtualRow, index) => {
    const originalRow = rows[virtualRow.index];
    let parity = virtualRow.index % 2 === 0 ? "even" : "odd";
    const isDraggedRow = draggingRowId === void 0 ? false : originalRow.id === draggingRowId || originalRow.id === draggingRowIdExpanded;
    const detailsParity = originalRow[DATA_TABLE_DETAILS_ROW_PARITY];
    if (detailsParity !== void 0) {
      parity = detailsParity;
    }
    if (originalRow[DATA_TABLE_DETAILS_ROW_IDENTIFIER] === true) {
      return /* @__PURE__ */ jsx(
        RowDetailsElement,
        {
          originalRow,
          virtualRow,
          virtualItemIndex: index,
          rowSeparation,
          rowParity: parity,
          isLastRowOnPage: virtualRow.index === rows.length - 1,
          rowMeasure,
          isDraggedRow,
          children: table.options.rowDetailsTemplate?.({
            row: originalRow.original
          })
        },
        originalRow.id
      );
    }
    return /* @__PURE__ */ jsx(
      DataTableRow,
      {
        row: originalRow,
        virtualRow,
        virtualItemIndex: index,
        rowCount: rows.length,
        virtualColumns,
        isDraggedRow,
        isInDragOverlay: false,
        rowParity: parity,
        rowMeasure,
        columnMeasure: (
          // For columns, only measure the first rendered row.
          // All other rows will have the same column widths.
          firstRenderedRowIndex === virtualRow.index ? columnMeasure : void 0
        ),
        scrollToIndex: rowVirtualizer.scrollToIndex
      },
      `row-${virtualRow.key}`
    );
  });
  const bodyContent = table.options.enableRowOrderDragAndDrop ? /* @__PURE__ */ jsx(
    RowOrderingDragAndDropContext,
    {
      rows,
      virtualColumns,
      rowSeparation,
      children: rowElements
    }
  ) : (
    // If row order drag and drop is not enabled then don't use the RowOrderingDragAndDropContext
    // eslint-disable-next-line react/jsx-no-useless-fragment
    /* @__PURE__ */ jsx(Fragment, { children: rowElements })
  );
  return /* @__PURE__ */ jsx(
    "div",
    {
      role: "rowgroup",
      "data-rowgroup": "body",
      className: clsx(styles.body, {
        [styles.loading]: table.getState().loading
      }),
      children: bodyContent
    }
  );
}
DataTableBody.displayName = "DataTableBody";
export {
  DataTableBody
};
