// @dynatrace/strato-components/esm/tables/DataTable/components/DataTableHeader.js
import { Fragment, jsx } from "react/jsx-runtime";
import { useEffect, useRef } from "react";
import * as headerStyles from "./DataTableHeader.sty.js";
import { DataTableHeaderCell } from "./DataTableHeaderCell.js";
import {
  DataTableIsHeaderContextProvider,
  useDataTable
} from "../contexts/DataTableContext.js";
import { ColumnOrderDragAndDropContext } from "../features/ColumnOrder/ColumnOrderDragAndDropContext.js";
function DataTableHeader() {
  const table = useDataTable();
  const headerGroups = table?.getHeaderGroups();
  const headerRef = useRef(null);
  useEffect(() => {
    if (!headerRef.current) {
      return;
    }
    const resizeObserver = new ResizeObserver((entries) => {
      const height = entries.at(0)?.borderBoxSize?.at(0)?.blockSize;
      table.setObservedHeaderHeight(height ?? 0);
    });
    resizeObserver.observe(headerRef.current);
    return () => {
      resizeObserver.disconnect();
    };
  }, [table]);
  const headerRows = headerGroups.map((headerGroup, rowIndex, rows) => {
    const headers = table.getOrderedHeaderGroups()[rowIndex];
    const headerCells = headers.map((header, columnIndex) => {
      return /* @__PURE__ */ jsx(
        DataTableHeaderCell,
        {
          header,
          headers,
          headerGroupId: headerGroup.id,
          rowIndex,
          columnIndex,
          lastRowIndex: rows.length - 1
        },
        header.id
      );
    });
    return /* @__PURE__ */ jsx("div", { role: "row", className: headerStyles.headerRow, children: headerCells }, headerGroup.id);
  });
  const headerContent = table.options.enableColumnOrdering ? /* @__PURE__ */ jsx(ColumnOrderDragAndDropContext, { children: headerRows }) : (
    // If column ordering is not enabled, don't use the ColumnOrderDragAndDropContext
    // eslint-disable-next-line react/jsx-no-useless-fragment
    /* @__PURE__ */ jsx(Fragment, { children: headerRows })
  );
  return /* @__PURE__ */ jsx(DataTableIsHeaderContextProvider, { children: /* @__PURE__ */ jsx(
    "div",
    {
      ref: headerRef,
      role: "rowgroup",
      "data-rowgroup": "header",
      className: headerStyles.header,
      children: headerContent
    }
  ) });
}
DataTableHeader.displayName = "DataTableHeader";
export {
  DataTableHeader
};
