// @dynatrace/strato-components/esm/tables/DataTable/components/DataTableVirtualizationContainer.js
import { jsx, jsxs } from "react/jsx-runtime";
import { useVirtualizer } from "@tanstack/react-virtual";
import clsx from "clsx";
import { debounce } from "lodash-es";
import {
  useCallback,
  useEffect,
  useMemo,
  useRef
} from "react";
import { DataTableContent } from "./DataTableContent.js";
import { DataTableStatePlaceholder } from "./DataTableStatePlaceholder.js";
import * as styles from "./DataTableVirtualizationContainer.sty.js";
import { useFontsUpdated } from "../../../core/hooks/useFontsUpdated.js";
import { devLog } from "../../../core/utils/logging.js";
import { getDataTableBorderWidth } from "../../utils/data-table-border-width.js";
import {
  DATA_TABLE_DEFAULT_COLUMN_WIDTH,
  DATA_TABLE_DEFAULT_MIN_COLUMN_WIDTH,
  DATA_TABLE_VIRTUALIZER_COLUMN_OVERSCAN,
  DATA_TABLE_VIRTUALIZER_ROW_OVERSCAN
} from "../constants.js";
import { addPinnedColumnsToVisibleRange } from "../features/ColumnPinning/column-pinning-utils.js";
import { isWidthExtendedConfig } from "../features/Layout/width-utils.js";
function DataTableVirtualizationContainer(props) {
  const { table, emptyState, fullHeight } = props;
  const parentRef = useRef(null);
  const { draggingRowInfo, draggedHeaderColumn, loading } = table.getState();
  const rows = table.getRowModel().rows;
  const visibleLeafColumns = table.getVisibleLeafColumns();
  const totalHeaderHeight = table.getTotalHeaderHeight();
  const visibleLeafColumnsWithoutBuiltin = table.getVisibleLeafColumnsWithoutBuiltin();
  const getUniqueRowKey = useCallback(
    (index) => {
      if (index >= rows.length) {
        devLog(`[Table] Invalid row index ${index} (>= ${rows.length})`);
        return -1;
      }
      return rows[index].id;
    },
    [rows]
  );
  const rowVirtualizer = useVirtualizer({
    count: rows?.length ?? 0,
    getScrollElement: () => parentRef.current,
    // Overscan of 5 was very ambitious. A single mouse scroll tick is usually
    // more than 5 condensed rows, so this is increased. In the future there could
    // be a possibility to make this dependent on the row density.
    overscan: DATA_TABLE_VIRTUALIZER_ROW_OVERSCAN,
    indexAttribute: "data-row",
    // reserves space for the sticky header in the scrollable area, so the first rows are not hidden behind the header when scrolled to the top
    paddingStart: totalHeaderHeight,
    // ensures that scrollToIndex scrolls the item to the correct position, taking into account the reserved space for the header
    scrollPaddingStart: totalHeaderHeight,
    estimateSize: (index) => table.estimateRowHeight({ index }),
    measureElement: (element) => element.getBoundingClientRect().height,
    // Default wait until the scrollEnd is triggered and it starts updating.
    // https://tanstack.com/virtual/latest/docs/api/virtualizer#isscrollingresetdelay
    // Reducing it from the default 150ms seems to improve perceived scroll performance
    isScrollingResetDelay: 20,
    getItemKey: getUniqueRowKey,
    // reset scrollElement observers when the data in the table is changed
    // fixes the issue with Firefox browser not resetting the scrollTop position of the virtualizer when the table is scrolled and the data is changed
    // APPDEV-15467
    enabled: rows.length > 0
  });
  const pinnedLeftColumns = table.getLeftVisibleLeafColumns();
  const pinnedRightColumns = table.getRightVisibleLeafColumns();
  const pinnedLeftIndexes = useMemo(
    () => Array.from({ length: pinnedLeftColumns.length }, (_, i) => i),
    [pinnedLeftColumns.length]
  );
  const pinnedRightIndexes = useMemo(() => {
    const pinnedRightStartIndex = visibleLeafColumns.length - pinnedRightColumns.length;
    return Array.from(
      { length: pinnedRightColumns.length },
      (_, i) => pinnedRightStartIndex + i
    );
  }, [visibleLeafColumns.length, pinnedRightColumns.length]);
  const columnRangeExtractor = useCallback(
    (range) => addPinnedColumnsToVisibleRange(
      range,
      pinnedLeftIndexes,
      pinnedRightIndexes
    ),
    [pinnedLeftIndexes, pinnedRightIndexes]
  );
  const orderedVisibleLeafColumns = table.getOrderedVisibleLeafColumns();
  const getUniqueColumnKey = useCallback(
    (index) => orderedVisibleLeafColumns[index].id,
    [orderedVisibleLeafColumns]
  );
  const widthConfigsById = useMemo(() => {
    return visibleLeafColumnsWithoutBuiltin.reduce(
      (acc, column) => {
        const { id, columnDef } = column;
        const {
          width,
          minWidth = DATA_TABLE_DEFAULT_MIN_COLUMN_WIDTH,
          maxWidth
        } = columnDef;
        const isExtendedConfig = isWidthExtendedConfig(width);
        const isContentOrAutoWidth = isExtendedConfig || width === "content" || width === "auto";
        const maxAutoWidth = isExtendedConfig ? width.maxWidth : void 0;
        const finalMaxWidth = maxWidth !== void 0 && maxAutoWidth !== void 0 ? Math.min(maxWidth, maxAutoWidth) : maxWidth ?? maxAutoWidth;
        acc[id] = {
          minWidth,
          maxWidth: finalMaxWidth,
          isContentOrAutoWidth
        };
        return acc;
      },
      {}
    );
  }, [visibleLeafColumnsWithoutBuiltin]);
  const columnMeasure = useCallback(
    (element) => {
      const measuredWidth = element.getBoundingClientRect().width;
      const columnId = element instanceof HTMLElement ? element.dataset.columnid : void 0;
      const previousWidth = columnId ? table.getMeasuredColumnWidth(columnId) : void 0;
      if (measuredWidth === 0 || !columnId) {
        return previousWidth ?? 0;
      }
      const widthConfiguration = widthConfigsById[columnId];
      if (table.getState().columnSizing[columnId] !== void 0 || table.getState().columnSizingInfo.isResizingColumn === columnId || !widthConfiguration) {
        table.setMeasuredColumnWidth(columnId, measuredWidth);
        return measuredWidth;
      }
      const { minWidth, maxWidth, isContentOrAutoWidth } = widthConfiguration;
      const roundedWidth = isContentOrAutoWidth ? Math.ceil(measuredWidth * 10) / 10 : measuredWidth;
      const clampedWidth = Math.max(
        minWidth ?? 0,
        Math.min(maxWidth ?? Infinity, roundedWidth)
      );
      if (previousWidth !== void 0 && Math.abs(clampedWidth - previousWidth) <= 0.1) {
        return previousWidth;
      }
      table.setMeasuredColumnWidth(columnId, clampedWidth);
      return clampedWidth;
    },
    [table, widthConfigsById]
  );
  const columnVirtualizer = useVirtualizer({
    count: visibleLeafColumns.length,
    getScrollElement: () => parentRef.current,
    indexAttribute: "data-column",
    estimateSize: () => DATA_TABLE_DEFAULT_COLUMN_WIDTH,
    rangeExtractor: columnRangeExtractor,
    overscan: DATA_TABLE_VIRTUALIZER_COLUMN_OVERSCAN,
    horizontal: true,
    getItemKey: getUniqueColumnKey,
    measureElement: columnMeasure
  });
  const hasVisibleColumns = visibleLeafColumnsWithoutBuiltin.length > 0;
  const { headers, contained, rowSeparation } = table.getVisualVariantOptions();
  const estimatedRowHeight = table.estimateRowHeight();
  const containerStyle = useMemo(() => {
    let minHeight = void 0;
    if (totalHeaderHeight === 0) {
      minHeight = headers === "hidden" ? estimatedRowHeight : estimatedRowHeight * 2 + getDataTableBorderWidth();
    } else {
      minHeight = totalHeaderHeight + estimatedRowHeight;
    }
    if (contained) {
      minHeight += 2 * getDataTableBorderWidth();
    }
    const baseStyle = {
      minHeight
    };
    if (fullHeight) {
      baseStyle.height = "100%";
    }
    return baseStyle;
  }, [totalHeaderHeight, contained, headers, estimatedRowHeight, fullHeight]);
  const scrollRect = rowVirtualizer.scrollRect;
  const scrollElement = rowVirtualizer.scrollElement;
  const scrollWidth = scrollElement?.scrollWidth;
  const scrollHeight = scrollElement?.scrollHeight;
  const scrollDirection = useMemo(() => {
    if (scrollRect?.width && scrollRect?.height && scrollWidth !== void 0 && scrollHeight !== void 0) {
      const hasScrollbarX = scrollRect.width < scrollWidth;
      const hasScrollbarY = scrollRect.height < scrollHeight;
      if (hasScrollbarX && hasScrollbarY) {
        return "xy";
      } else if (hasScrollbarX) {
        return "x";
      } else if (hasScrollbarY) {
        return "y";
      }
    }
    return false;
  }, [scrollRect?.width, scrollRect?.height, scrollWidth, scrollHeight]);
  const fontUpdateQueued = useRef(false);
  useFontsUpdated(() => {
    if (fontUpdateQueued.current) {
      return;
    }
    fontUpdateQueued.current = true;
    const controller = new AbortController();
    Promise.race([
      // If the animations have not stopped after a timeout period,
      // we can remeasure nevertheless.
      new Promise(
        (resolve) => setTimeout(() => {
          resolve();
        }, 1e3)
      ),
      // If the animations subside within a reasonable interval, we
      // can also remeasure.
      new Promise((resolve) => {
        document.addEventListener(
          "animationend",
          debounce(() => {
            resolve();
          }, 200),
          { signal: controller.signal }
        );
      })
    ]).then(() => {
      table.remeasureAllColumns();
      controller.abort();
      fontUpdateQueued.current = false;
    });
  });
  useEffect(() => {
    const currentColumnIds = new Set(
      table.getAllLeafColumns().map((col) => col.id)
    );
    const removedIds = [];
    for (const id of table.getMeasuredColumnWidths().keys()) {
      if (!currentColumnIds.has(id)) {
        removedIds.push(id);
      }
    }
    if (removedIds.length > 0) {
      table.resetMeasuredColumnWidths(removedIds);
    }
  }, [table, table.options.columns]);
  useEffect(() => {
    const parent = parentRef.current;
    const currentOffset = rowVirtualizer.scrollOffset;
    const abort = new AbortController();
    if (draggedHeaderColumn) {
      parent?.addEventListener(
        "scroll",
        () => rowVirtualizer.scrollToOffset(currentOffset ?? 0),
        { signal: abort.signal }
      );
    }
    return () => {
      abort.abort();
    };
  }, [draggedHeaderColumn, rowVirtualizer]);
  useEffect(() => {
    const parent = parentRef.current;
    const currentOffset = columnVirtualizer.scrollOffset;
    const abort = new AbortController();
    if (draggingRowInfo) {
      parent?.addEventListener(
        "scroll",
        () => columnVirtualizer.scrollToOffset(currentOffset ?? 0),
        { signal: abort.signal }
      );
    }
    return () => {
      abort.abort();
    };
  }, [columnVirtualizer, draggingRowInfo]);
  table.registerVirtualizers(rowVirtualizer, columnVirtualizer);
  return /* @__PURE__ */ jsxs(
    "div",
    {
      ref: parentRef,
      className: clsx(
        "strato-table-virtualization-container",
        styles.virtualizationContainer({
          contained,
          rowSeparation,
          headers,
          loading,
          hasVisibleColumns,
          scrollDirection
        })
      ),
      style: containerStyle,
      children: [
        /* @__PURE__ */ jsx(
          DataTableContent,
          {
            rowVirtualizer,
            columnVirtualizer
          }
        ),
        /* @__PURE__ */ jsx(
          DataTableStatePlaceholder,
          {
            isColumnsEmpty: columnVirtualizer.options.count === 0,
            isDataEmpty: rowVirtualizer.options.count === 0,
            allColumnsHidden: !hasVisibleColumns,
            emptyStateSlot: emptyState
          }
        )
      ]
    }
  );
}
export {
  DataTableVirtualizationContainer
};
