// @dynatrace/strato-components/esm/tables/DataTable/components/DataTableRow.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { useSortable } from "@dnd-kit/sortable";
import clsx from "clsx";
import { useRef } from "react";
import { isTargetInteractive } from "../../utils/is-target-interactive.js";
import {
  DataTableRowContextProvider,
  useDataTable
} from "../contexts/DataTableContext.js";
import * as tableStyles from "../DataTable.sty.js";
import { DataTablePinningShadowCells } from "./ColumnPinning/DataTablePinningShadowCells.js";
import { DataTableCell } from "./DataTableCell.js";
import * as styles from "./DataTableRow.sty.js";
import { rowDetailsExpandedRowCell } from "../features/RowDetails/RowDetails.sty.js";
import { useRowInteractivityProps } from "../features/RowInteractivity/RowInteractivity.js";
import { SubRowCellContextProvider } from "../features/SubRows/SubRowCellContext.js";
const ROW_LONG_PRESS_TIMEOUT_DURATION = 500;
function DataTableRow(props) {
  const {
    row,
    virtualRow,
    rowCount,
    virtualColumns,
    isDraggedRow,
    isInDragOverlay,
    virtualItemIndex,
    rowParity,
    rowMeasure,
    columnMeasure,
    scrollToIndex: scrollToIndexProp
  } = props;
  const table = useDataTable();
  const virtualRowIndex = virtualRow.index;
  const hasSubRows = table.options.enableExpanding === true;
  const subRowIndicatorColumnId = hasSubRows ? table.getSubRowIndicatorColumnId() : null;
  const subRowDepth = hasSubRows ? row.getSubRowDepth() : 0;
  const thresholdPill = row.getThresholdPill();
  const tableState = table.getState();
  const floatingMenuInfoRowId = tableState.floatingMenuInfo?.rowId;
  const floatingMenuInfoColumnId = tableState.floatingMenuInfo?.columnId;
  const scrollToIndex = scrollToIndexProp ?? (() => void 0);
  const { rowInteractivityClassName, rowInteractivityProps } = useRowInteractivityProps(row, scrollToIndex);
  const rowLink = row.getLinkForRow();
  const rowLinkRef = useRef(null);
  const { setNodeRef } = useSortable({
    id: row.id,
    data: {
      row,
      rowIndex: virtualRowIndex,
      virtualRow
    }
  });
  const isLongPress = useRef(false);
  const longPressTimeout = useRef(null);
  const handleOpenRowLinkOnEnter = (event) => {
    if (!rowLink) {
      return;
    }
    const isRowEnter = event.target.getAttribute("role") === "row" && event.key === "Enter";
    if (isRowEnter) {
      rowLinkRef.current?.click();
    }
  };
  const handleClickOnRowLink = (event) => {
    const { target } = event;
    if (isLongPress.current || !(target instanceof HTMLElement) && !(target instanceof SVGElement)) {
      return;
    }
    const row2 = target.closest('[role="row"]');
    if (row2 === null || isTargetInteractive(row2, target)) {
      return;
    }
    const nativeEvent = event.nativeEvent;
    const newEvent = new MouseEvent("click", {
      ...event.nativeEvent,
      // For some reason the Ctrl and Meta key as well as the button key have to be manually set,
      // even though the native event already contains the correct values
      ctrlKey: nativeEvent.ctrlKey === true,
      metaKey: nativeEvent.metaKey === true,
      button: nativeEvent.button
    });
    rowLinkRef.current?.dispatchEvent(newEvent);
  };
  const leftPinnedColumns = table.getLeftVisibleLeafColumns();
  const lastLeftPinnedColumn = leftPinnedColumns.at(-1);
  const lastLeftPinnedVirtualColumnIndex = leftPinnedColumns.length - 1;
  const rightPinnedColumns = table.getRightVisibleLeafColumns();
  const firstRightPinnedColumn = rightPinnedColumns.at(0);
  const firstRightPinnedVirtualColumnIndex = virtualColumns[virtualColumns.length - rightPinnedColumns.length]?.index;
  const gridRowPosition = row.getGridRowPosition(virtualItemIndex);
  const visibleCells = row.getVisibleCells();
  let linkHandlingEvents = {};
  if (rowLink) {
    linkHandlingEvents = {
      onPointerUp: handleClickOnRowLink,
      onKeyUp: handleOpenRowLinkOnEnter
    };
  }
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsxs(
      "div",
      {
        role: "row",
        ref: rowMeasure,
        "data-rowid": row.id,
        "data-row": virtualRowIndex,
        className: clsx(
          styles.row({
            rowSeparation: table.getVisualVariantOptions().rowSeparation,
            rowParity,
            isDraggedRow
          }),
          {
            [tableStyles.firstBodyRow]: virtualRowIndex === 0,
            [tableStyles.lastBodyRow]: virtualRowIndex === rowCount - 1,
            [`${rowDetailsExpandedRowCell}`]: row.getIsDetailsOpen(),
            [rowInteractivityClassName]: !isInDragOverlay
          }
        ),
        style: {
          ...gridRowPosition
        },
        onTouchStart: () => {
          longPressTimeout.current = setTimeout(() => {
            isLongPress.current = true;
          }, ROW_LONG_PRESS_TIMEOUT_DURATION);
        },
        onTouchEnd: () => {
          if (longPressTimeout.current !== null) {
            clearTimeout(longPressTimeout.current);
          }
          isLongPress.current = false;
        },
        ...!isInDragOverlay && {
          ...rowInteractivityProps,
          ...linkHandlingEvents
        },
        children: [
          /* @__PURE__ */ jsx(DataTableRowContextProvider, { value: row, children: virtualColumns.map((virtualCell) => {
            const virtualCellIndex = virtualCell.index;
            const cell = visibleCells[virtualCellIndex];
            const column = cell.column;
            const hasActiveFloatingMenu = floatingMenuInfoRowId === row.id && floatingMenuInfoColumnId === column.id;
            return /* @__PURE__ */ jsx(
              SubRowCellContextProvider,
              {
                isSubRowIndicatorColumn: hasSubRows && subRowIndicatorColumnId === column.id,
                cell,
                children: /* @__PURE__ */ jsx(
                  DataTableCell,
                  {
                    cell,
                    virtualColumnIndex: virtualCellIndex,
                    virtualRowIndex,
                    isInDragOverlay,
                    hasActiveFloatingMenu,
                    columnMeasure,
                    dragAndDropRef: setNodeRef,
                    subRowDepth,
                    thresholdPill,
                    rowData: row.original,
                    cellDeps: column.columnDef?.meta?.cellDeps?.(cell).toString()
                  }
                )
              },
              `${virtualRow.key}-${virtualCell.key}`
            );
          }) }),
          rowLink && // this link element solely holds the link for interactive rows and is not visible to the user
          // it's not directly clickable; instead, the link is opened by interactions with the row (click, enter)
          // eslint-disable-next-line jsx-a11y/anchor-has-content
          /* @__PURE__ */ jsx(
            "a",
            {
              ref: rowLinkRef,
              href: rowLink,
              onPointerDown: (event) => event.stopPropagation(),
              style: { display: "none" }
            }
          )
        ]
      }
    ),
    !isDraggedRow && /* @__PURE__ */ jsx(
      DataTablePinningShadowCells,
      {
        lastLeftPinnedColumn: {
          column: lastLeftPinnedColumn,
          index: lastLeftPinnedVirtualColumnIndex
        },
        firstRightPinnedColumn: {
          column: firstRightPinnedColumn,
          index: firstRightPinnedVirtualColumnIndex
        },
        gridRowPosition,
        rowId: row.id
      }
    )
  ] });
}
export {
  DataTableRow
};
