// @dynatrace/strato-components/esm/tables/DataTable/components/DataTableEmptyState.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import { forwardRef } from "react";
import { useIntl } from "react-intl";
import * as styles from "./DataTableEmptyState.sty.js";
import { Text } from "../../../typography/text/Text.js";
import { useDataTable } from "../contexts/DataTableContext.js";
import { dataTableMessages } from "../messages.js";
function _isEmptyState(child) {
  return child?.type === DataTableEmptyState;
}
const DataTableEmptyState = /* @__PURE__ */ forwardRef(
  ({
    children,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    ...remainingProps
  }, forwardedRef) => {
    const intl = useIntl();
    const table = useDataTable();
    if (children && typeof children !== "string") {
      return /* @__PURE__ */ jsx(
        "div",
        {
          ref: forwardedRef,
          className: consumerClassName,
          style: consumerStyle,
          "data-testid": dataTestId,
          ...remainingProps,
          children
        }
      );
    }
    const isColumnsEmpty = table.getVisibleLeafColumnsWithoutBuiltin().length === 0;
    const isDataEmpty = table.getRowModel().rows.length === 0;
    const fallbackMessage = isDataEmpty ? intl.formatMessage(dataTableMessages.noDataAvailable) : intl.formatMessage(dataTableMessages.noColumnsAvailable);
    return /* @__PURE__ */ jsx(
      Text,
      {
        ref: forwardedRef,
        className: clsx([
          styles.emptyState({ minWidth: isColumnsEmpty }),
          consumerClassName
        ]),
        style: consumerStyle,
        "data-testid": dataTestId,
        ...remainingProps,
        children: children || fallbackMessage
      }
    );
  }
);
DataTableEmptyState.displayName = "EmptyState";
export {
  DataTableEmptyState,
  _isEmptyState
};
