// @dynatrace/strato-components/esm/tables/DataTable/components/DataTableCell.js
import { jsx, jsxs } from "react/jsx-runtime";
import { flexRender } from "@tanstack/react-table";
import { assignInlineVars } from "@vanilla-extract/dynamic";
import clsx from "clsx";
import {
  memo,
  useContext,
  useState
} from "react";
import { DATA_TABLE_DEFAULT_HORIZONTAL_ALIGNMENT } from "../constants.js";
import * as alignmentStyles from "./Alignment.sty.js";
import * as cellStyles from "./DataTableCell.sty.js";
import { useMergeRefs } from "../../../core/hooks/useMergeRefs.js";
import {
  DataTableColumnContextProvider,
  useDataTable,
  useDataTableContent
} from "../contexts/DataTableContext.js";
import * as tableStyles from "../DataTable.sty.js";
import { SPACER_COLUMN } from "../features/Layout/Layout.js";
import { isRowDetailsColumnDef } from "../features/RowDetails/RowDetails.js";
import { isRowDragColumnDef } from "../features/RowOrder/row-order-column-utils.js";
import { isRowSelectionColumnDef } from "../features/RowSelection/RowSelection.js";
import { SubrowCellContext } from "../features/SubRows/SubRowCellContext.js";
import { SubRowsCellExpander } from "../features/SubRows/SubRowExpanders.js";
import * as thresholdStyles from "../features/Thresholds/Thresholds.sty.js";
import { InlineUserActionsMenu } from "../features/UserActions/InlineUserActionsMenu.js";
import { ROW_ACTIONS_COLUMN } from "../features/UserActions/RowActions.js";
function InternalMemoizedCell(props) {
  const {
    rowId,
    virtualRowIndex,
    columnId,
    virtualColumnIndex,
    className,
    columnMeasure,
    dragAndDropRef,
    pillColor,
    enableExpanding,
    hasCustomCellRenderer,
    isSubRowIndicatorColumn,
    columnType,
    isLineWrapped,
    children,
    canCellAction,
    subRowDepth,
    leftPosition,
    rightPosition,
    cellActionAriaLabel,
    thresholdColor,
    thresholdBackgroundColor
  } = props;
  const [hovered, setHovered] = useState(false);
  const mergedRef = useMergeRefs([
    dragAndDropRef,
    columnMeasure
  ]);
  return /* @__PURE__ */ jsxs(
    "div",
    {
      role: columnId === SPACER_COLUMN.id ? "presentation" : "cell",
      ref: mergedRef,
      "data-index": `row-${virtualRowIndex}-column-${virtualColumnIndex}`,
      "data-columnid": columnId,
      "data-column": virtualColumnIndex,
      "data-columntype": columnType,
      "data-subrow-depth": subRowDepth,
      className,
      onMouseEnter: () => canCellAction && setHovered(true),
      onMouseLeave: () => canCellAction && setHovered(false),
      style: {
        ...assignInlineVars({
          [thresholdStyles.thresholdPillColor]: pillColor,
          [cellStyles.leftPositionVar]: leftPosition,
          [cellStyles.rightPositionVar]: rightPosition
        }),
        gridColumnStart: (virtualColumnIndex ?? 0) + 1,
        gridColumnEnd: "span 1",
        gridRowStart: 1,
        gridRowEnd: "span 1",
        color: thresholdColor,
        backgroundColor: thresholdBackgroundColor,
        minWidth: columnId === ROW_ACTIONS_COLUMN.id ? "max-content" : void 0
      },
      children: [
        enableExpanding === true && !hasCustomCellRenderer && isSubRowIndicatorColumn === true ? /* @__PURE__ */ jsx(SubRowsCellExpander, { columnType }) : null,
        !hasCustomCellRenderer && columnType !== "sparkline" && columnType !== "meterbar" && columnType !== "log-content" ? /* @__PURE__ */ jsx(
          "div",
          {
            className: clsx(
              cellStyles.innerCell({
                lineWrap: isLineWrapped
              })
            ),
            children
          }
        ) : children,
        canCellAction && /* @__PURE__ */ jsx(
          InlineUserActionsMenu,
          {
            headerGroupOrRowId: rowId,
            columnId,
            type: "cell",
            "aria-label": cellActionAriaLabel,
            "data-column": virtualColumnIndex,
            hovered
          }
        )
      ]
    }
  );
}
function arePropsEqual(prevProps, nextProps) {
  if (nextProps.hasCustomCellRenderer && nextProps.cellDeps === void 0) {
    return false;
  }
  const { children: _prevChildren, ...prevRest } = prevProps;
  const { children: _nextChildren, ...nextRest } = nextProps;
  const keys = Object.keys(nextRest);
  return keys.every((key) => prevRest[key] === nextRest[key]);
}
const DataTableCellMemo = memo(InternalMemoizedCell, arePropsEqual);
function DataTableCell(props) {
  const {
    cell,
    virtualColumnIndex,
    virtualRowIndex,
    isInDragOverlay,
    dragAndDropRef,
    columnMeasure,
    hasActiveFloatingMenu,
    subRowDepth,
    thresholdPill,
    rowData,
    cellDeps
  } = props;
  const table = useDataTable();
  const { cellActionAriaLabel } = useDataTableContent();
  const subrowContext = useContext(SubrowCellContext);
  const { row, column, getThresholdColors, getContext } = cell;
  const visualVariantOptions = table.getVisualVariantOptions();
  const isLineWrapped = column.getIsLineWrapped();
  const pillColor = virtualColumnIndex === 0 ? thresholdPill : null;
  const thresholdCellColors = getThresholdColors();
  const columnId = column.id;
  const columnDef = column.columnDef;
  const hasCustomCellRenderer = columnDef.meta?.hasCustomCellRenderer;
  const isFirstColumn = virtualColumnIndex === 0;
  const isPrefixColumn = isRowSelectionColumnDef(columnDef) || isRowDetailsColumnDef(columnDef) || isRowDragColumnDef(columnDef);
  const pinningState = column.getIsPinned();
  const leftPinnedPosition = pinningState === "left" ? column.getLeftPinnedPosition() : void 0;
  const rightPinnedPosition = pinningState === "right" ? column.getRightPinnedPosition() : void 0;
  const pinnedColumnBg = pinningState && visualVariantOptions.rowSeparation !== "zebraStripes" && (visualVariantOptions.contained ? "contained" : "uncontained");
  return /* @__PURE__ */ jsx(
    DataTableCellMemo,
    {
      rowId: row.id,
      virtualRowIndex: isInDragOverlay ? void 0 : virtualRowIndex,
      virtualColumnIndex,
      columnId,
      subRowDepth,
      canCellAction: column.getCanCellActions(),
      columnMeasure,
      dragAndDropRef: !isInDragOverlay && isFirstColumn ? dragAndDropRef : void 0,
      pillColor,
      thresholdColor: thresholdCellColors.color,
      thresholdBackgroundColor: thresholdCellColors.backgroundColor,
      enableExpanding: table.options.enableExpanding,
      leftPosition: leftPinnedPosition !== void 0 ? `${leftPinnedPosition}px` : void 0,
      rightPosition: rightPinnedPosition !== void 0 ? `${rightPinnedPosition}px` : void 0,
      hasCustomCellRenderer,
      isSubRowIndicatorColumn: subrowContext.isSubRowIndicatorColumn,
      columnType: columnDef.columnType,
      isLineWrapped,
      cellActionAriaLabel,
      className: clsx(
        cellStyles.outerCell({
          lineWrap: isLineWrapped,
          fontStyle: column.getFontStyle(),
          rowDensity: hasCustomCellRenderer ? "customCellRenderer" : visualVariantOptions.rowDensity,
          interactiveRows: table.options.enableRowInteractivity,
          verticalDividers: visualVariantOptions.verticalDividers,
          pinnedColumn: pinningState,
          background: pinnedColumnBg
        }),
        alignmentStyles.horizontalAlignment[columnDef.alignment ?? DATA_TABLE_DEFAULT_HORIZONTAL_ALIGNMENT],
        alignmentStyles.verticalAlignment[table.getBodyVerticalAlignment()],
        {
          [thresholdStyles.thresholdPill]: pillColor !== void 0 && pillColor !== null,
          [tableStyles.prefixColumn]: isPrefixColumn,
          [tableStyles.pinnedCellInDragOverlay]: pinningState && isInDragOverlay,
          [cellStyles.activeInteraction]: hasActiveFloatingMenu
        }
      ),
      rowData,
      tableColumns: table.options.columns,
      cellDeps,
      children: hasCustomCellRenderer ? /* @__PURE__ */ jsx(
        DataTableColumnContextProvider,
        {
          value: column,
          children: flexRender(column.columnDef.cell, getContext())
        }
      ) : flexRender(column.columnDef.cell, getContext())
    }
  );
}
export {
  DataTableCell
};
