// @dynatrace/strato-components/esm/tables/DataTable/components/DataTableContent.js
import { jsx, jsxs } from "react/jsx-runtime";
import { useIntl } from "react-intl";
import { DataTableBody } from "./DataTableBody.js";
import { DataTableHeader } from "./DataTableHeader.js";
import {
  DataTableContentContextProvider,
  useDataTable
} from "../contexts/DataTableContext.js";
import { CellActionFloatingMenu } from "../features/UserActions/FloatingMenu.js";
import { dataTableMessages } from "../messages.js";
import { DataTableScrollShadow } from "./DataTableScrollShadow.js";
import { getDataTableBorderWidth } from "../../utils/data-table-border-width.js";
const DataTableContent = (props) => {
  const intl = useIntl();
  const table = useDataTable();
  const visibleLeafColumnsWithoutBuiltin = table.getVisibleLeafColumnsWithoutBuiltin();
  const hasVisibleColumns = visibleLeafColumnsWithoutBuiltin.length > 0;
  if (!hasVisibleColumns) {
    return null;
  }
  const { rowVirtualizer, columnVirtualizer } = props;
  const rows = table.getRowModel().rows;
  const draggingRow = table.getState().draggingRowInfo;
  const isDraggingRow = !!draggingRow;
  const isDataEmpty = rowVirtualizer.options.count === 0;
  const isColumnsEmpty = columnVirtualizer.options.count === 0;
  const isEmpty = isDataEmpty || isColumnsEmpty;
  const scrollOffset = columnVirtualizer.scrollOffset ?? 0;
  const scrollElement = columnVirtualizer.scrollElement;
  const tableViewportWidth = scrollElement?.clientWidth ?? 0;
  const scrollElementWidth = scrollElement?.scrollWidth ?? 0;
  const visualVariantOptions = table.getVisualVariantOptions();
  const hasOverflowLeft = scrollOffset > 0;
  const gridHeight = isEmpty ? "auto" : rowVirtualizer.getTotalSize();
  const borderWidth = visualVariantOptions.contained ? getDataTableBorderWidth() : 0;
  const hasOverflowRight = scrollElementWidth - scrollOffset - borderWidth > tableViewportWidth;
  const cellActionAriaLabel = intl.formatMessage(
    dataTableMessages.cellActionLabel
  );
  return /* @__PURE__ */ jsxs(
    "div",
    {
      role: "grid",
      style: {
        display: "grid",
        height: gridHeight,
        gridTemplateColumns: table.getGridTemplateColumns(),
        gridTemplateRows: table.getGridTemplateRows()
      },
      onMouseOver: (event) => {
        if (!isDraggingRow) {
          table.handleInteractiveRowMouseOver(event);
        }
      },
      onMouseLeave: (event) => {
        table.handleInteractiveRowMouseLeave(event);
      },
      onMouseDown: (event) => {
        if (
          // main button (left button)
          event.button === 0 || // auxiliary button (wheel/middle button)
          event.button === 1
        ) {
          table.handleInteractiveRowClick(event);
        }
      },
      onContextMenu: (event) => {
        if (event.target instanceof Element && event.target.closest("a")) {
          return;
        }
        table.getUserActionFloatingMenuContextMenuHandler()(event);
      },
      children: [
        /* @__PURE__ */ jsxs(
          DataTableContentContextProvider,
          {
            value: {
              hasOverflowLeft,
              hasOverflowRight,
              cellActionAriaLabel
            },
            children: [
              visualVariantOptions.headers === "default" && /* @__PURE__ */ jsx(DataTableHeader, {}),
              /* @__PURE__ */ jsx(
                DataTableBody,
                {
                  rows,
                  rowVirtualizer,
                  columnVirtualizer
                }
              ),
              (table.options.enableCellActions || table.options.enableColumnActions || table.options.enableSorting) && /* @__PURE__ */ jsx(CellActionFloatingMenu, {})
            ]
          }
        ),
        /* @__PURE__ */ jsx(
          DataTableScrollShadow,
          {
            hasOverflowLeft,
            hasOverflowRight
          }
        )
      ]
    }
  );
};
export {
  DataTableContent
};
