// @dynatrace/strato-components/charts/core/components/wheel-zoom-preventer/WheelZoomPreventer.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var WheelZoomPreventer_exports = {};
__export(WheelZoomPreventer_exports, {
  WheelZoomPreventer: () => WheelZoomPreventer
});
module.exports = __toCommonJS(WheelZoomPreventer_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_usePreventScrollOnWheel = require("../../hooks/usePreventScrollOnWheel.js");
const WheelZoomPreventer = () => {
  const ref = (0, import_react.useRef)(null);
  (0, import_usePreventScrollOnWheel.usePreventScrollOnWheel)(ref);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      ref,
      style: {
        position: "relative",
        width: "100%",
        height: "100%"
      }
    }
  );
};
WheelZoomPreventer["displayName"] = "WheelZoomPreventer";
