// @dynatrace/strato-components/typography/list/List.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var List_exports = {};
__export(List_exports, {
  List: () => List
});
module.exports = __toCommonJS(List_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_List_css = require("./List.sty.js");
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useSafeForwardProps = require("../../core/hooks/useSafeForwardProps.js");
var import_Text = require("../text/Text.js");
const COMPONENT_NAME = "List";
function interleaveListItems(listItems) {
  const items = [];
  import_react.Children.forEach(listItems, (listItem) => {
    if ((0, import_react.isValidElement)(listItem)) {
      if (listItem.type === List) {
        if (items.length === 0) {
          items.push([listItem]);
        } else {
          items[items.length - 1].push(listItem);
        }
      } else {
        items.push([listItem]);
      }
    }
  });
  return items.map((values, index) => {
    const itemKey = `list-item-${getItemKey(values, index)}`;
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("li", { className: import_List_css.listItemCSS, children: values }, itemKey);
  });
}
function getItemKey(values, index) {
  const isArray = Array.isArray(values);
  const value = isArray ? values?.[0] : values;
  if ((0, import_react.isValidElement)(value)) {
    if (value.key) {
      return value.key;
    } else if (value.props && value.props.children) {
      return getItemKey(value.props.children, index);
    }
    return value;
  }
  if (value === void 0) {
    return index;
  }
  return value;
}
const List = /* @__PURE__ */ (0, import_react.forwardRef)(
  (props, ref) => {
    const {
      children,
      ordered = void 0,
      textStyle,
      fontStyle,
      start = 1,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      className: consumerClassName,
      style: consumerStyle,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
    const ariaLabelingProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(restProps);
    const forwardProps = (0, import_useSafeForwardProps.useSafeForwardProps)(restProps, ariaLabelingProps);
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Text.Text,
      {
        ref,
        as: ordered ? "ol" : "ul",
        ...ordered && {
          start
        },
        textStyle,
        fontStyle,
        className: (0, import_clsx.default)(import_List_css.listCSS, consumerClassName),
        style: consumerStyle,
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        ...ariaLabelingProps,
        ...behavioralTrackingProps,
        ...forwardProps,
        "data-dt-component": COMPONENT_NAME,
        children: interleaveListItems(children)
      }
    );
  }
);
List.displayName = COMPONENT_NAME;
