// @dynatrace/strato-components/esm/charts/histogram/components/displays/utils/datapoint-stacking.js
function getStackedKeyByUnit(bin, unit) {
  return bin.from.toString() + Math.sign(bin.value).toString() + bin.from.toString() + bin.to.toString() + //TODO: Pending to apply hash function from this PR: https://bitbucket.lab.dynatrace.org/projects/APPFW/repos/design-system/pull-requests/7188/overview?commentId=1144690
  (unit ?? "").toString();
}
function getCumulativeValue(mapStackedPoints, bin, unit, seriesId, value = 0) {
  const key = getStackedKeyByUnit(bin, unit);
  const stackedBin = mapStackedPoints.get(key);
  let binsCumulative = 0;
  let baseCumulative = 0;
  let cumulativeValue = value;
  if (stackedBin) {
    const {
      binsCumulative: prevBinsCumulative,
      baseCumulative: prevBaseCumulative,
      seriesId: prevSeriesId
    } = stackedBin;
    binsCumulative = prevBinsCumulative ?? 0;
    baseCumulative = prevBaseCumulative ?? 0;
    const baseCumulativeHasToChange = prevSeriesId !== seriesId;
    if (baseCumulativeHasToChange) {
      baseCumulative = binsCumulative;
    }
    cumulativeValue = baseCumulative + value;
  }
  binsCumulative = stackedBin?.seriesId !== seriesId ? cumulativeValue : Math.sign(value) > 0 ? Math.max(cumulativeValue, binsCumulative) : Math.min(cumulativeValue, binsCumulative);
  return {
    cumulativeValue,
    binsCumulative,
    baseCumulative,
    key
  };
}
function setCumulativeBins(mapStackedPoints, seriesId, key, binsCumulative, baseCumulative) {
  mapStackedPoints.set(key, {
    seriesId,
    binsCumulative,
    baseCumulative
  });
}
export {
  getCumulativeValue,
  getStackedKeyByUnit,
  setCumulativeBins
};
