// @dynatrace/strato-components/charts/tree-map/components/TreeMapInternal.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TreeMapInternal_exports = {};
__export(TreeMapInternal_exports, {
  TreeMapInternal: () => TreeMapInternal
});
module.exports = __toCommonJS(TreeMapInternal_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_hierarchy = require("@visx/hierarchy");
var import_react = require("react");
var import_TreeMapCanvas = require("./TreeMapCanvas.js");
var import_TreeMapCluster = require("./TreeMapCluster.js");
var import_Canvas = require("../../core/components/canvas/Canvas.js");
var import_CanvasContext = require("../../core/components/canvas/CanvasContext.js");
var import_useGraphSize = require("../../core/components/chart-layout/hooks/useGraphSize.js");
var import_constants = require("../constants.js");
var import_LabelsDisplayContext = require("../context/LabelsDisplayContext.js");
var import_TreeMapAggregationDataContext = require("../context/TreeMapAggregationDataContext.js");
var import_TreeMapHighlightingMapContext = require("../context/TreeMapHighlightingMapContext.js");
var import_selectors = require("../store/selectors.js");
var import_node_padding = require("../utils/node-padding.js");
var import_with_parent_label = require("../utils/with-parent-label.js");
const TreeMapInternal = (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const { children } = props;
    const { width, height } = (0, import_useGraphSize.useGraphSize)();
    const roundedWidth = Math.floor(width);
    const roundedHeight = Math.floor(height);
    const labelsDisplay = (0, import_LabelsDisplayContext.useLabelsDisplay)();
    const { dataForRendering } = (0, import_TreeMapAggregationDataContext.useTreeMapAggregationData)();
    const highlightingMap = (0, import_TreeMapHighlightingMapContext.useTreeMapHighlightingMap)();
    const hasParentLabels = (0, import_with_parent_label.withParentLabel)(labelsDisplay);
    const root = (0, import_react.useMemo)(() => (0, import_hierarchy.hierarchy)(dataForRendering), [dataForRendering]);
    const highlightedItem = (0, import_selectors.useHighlightedState)();
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { ref: forwardedRef, "aria-label": import_constants.TREEMAP_CHART_ARIA_LABEL, role: "img", children: [
      children,
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_CanvasContext.CanvasContext.Provider,
        {
          value: {
            shapeRefs: /* @__PURE__ */ new Map()
          },
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Canvas.Canvas, { width: roundedWidth, height: roundedHeight, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_hierarchy.Treemap,
            {
              root,
              size: [roundedWidth, roundedHeight],
              tile: import_hierarchy.treemapSquarify,
              padding: (node) => (0, import_node_padding.nodePadding)(node, hasParentLabels),
              paddingTop: hasParentLabels ? (node) => (0, import_node_padding.nodePaddingTop)(node) : void 0,
              paddingOuter: hasParentLabels ? (node) => (0, import_node_padding.nodePaddingOuter)(node) : void 0,
              round: true,
              children: (rootNode) => /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
                hasParentLabels && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  import_TreeMapCluster.TreeMapCluster,
                  {
                    treeMapData: rootNode,
                    highlightedItem,
                    highlightingMap
                  }
                ),
                /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  import_TreeMapCanvas.TreeMapCanvas,
                  {
                    treeMapData: rootNode,
                    highlightedItem,
                    highlightingMap
                  }
                )
              ] })
            }
          ) })
        }
      )
    ] });
  }
);
