// @dynatrace/strato-components/charts/tree-map/components/TreeMapChartLayout.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TreeMapChartLayout_exports = {};
__export(TreeMapChartLayout_exports, {
  TreeMapChartLayout: () => TreeMapChartLayout
});
module.exports = __toCommonJS(TreeMapChartLayout_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_TreeMapInternal = require("./TreeMapInternal.js");
var import_TreeMapLegendRenderer = require("./TreeMapLegend/TreeMapLegendRenderer.js");
var import_TreeMapOverlay = require("./TreeMapOverlay/TreeMapOverlay.js");
var import_useBehavioralTrackingProps = require("../../../core/hooks/useBehavioralTrackingProps.js");
var import_ChartLayout = require("../../core/components/chart-layout/ChartLayout.js");
var import_OverlayContext = require("../../core/components/overlay-tooltip/contexts/OverlayContext.js");
var import_TooltipContext = require("../../core/components/overlay-tooltip/contexts/TooltipContext.js");
var import_ChartLegendContext = require("../context/ChartLegendContext.js");
var import_TreeMapAggregationDataProvider = require("../providers/TreeMapAggregationDataProvider.js");
var import_TreeMapHighlightingMapProvider = require("../providers/TreeMapHighlightingMapProvider.js");
var import_TreeMapInternalDataProvider = require("../providers/TreeMapInternalDataProvider.js");
const TreeMapChartLayout = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    chartLayoutRef,
    width,
    height,
    emptyState,
    errorState,
    isEmpty,
    loading,
    className,
    style,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "data-testid": dataTestId,
    "data-dt-component": dataDtComponent
  } = props;
  const [behavioralTrackingProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(props);
  const {
    hidden: legendHidden,
    position: legendPosition,
    // eslint-disable-next-line @typescript-eslint/no-deprecated
    ratio,
    initialRatio,
    onRatioChange
  } = (0, import_ChartLegendContext.useChartLegend)();
  const legendRatio = initialRatio ?? ratio;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_ChartLayout.ChartLayout,
    {
      ...behavioralTrackingProps,
      ref: chartLayoutRef,
      chartWidth: width,
      chartHeight: height,
      emptyState,
      errorState,
      showLoader: loading,
      isEmpty,
      className,
      style,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "data-testid": dataTestId,
      "data-dt-component": dataDtComponent,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ChartLayout.ChartLayout.Graph, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TreeMapInternalDataProvider.TreeMapInternalDataProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TreeMapAggregationDataProvider.TreeMapAggregationDataProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TreeMapHighlightingMapProvider.TreeMapHighlightingMapProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayContext.OverlayProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TooltipContext.TooltipProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { style: { position: "relative" }, children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TreeMapInternal.TreeMapInternal, { ref: forwardedRef }),
          loading ? null : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TreeMapOverlay.TreeMapOverlay, {})
        ] }) }) }) }) }) }) }),
        !legendHidden && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_ChartLayout.ChartLayout.Legend,
          {
            position: legendPosition,
            initialRatio: legendRatio,
            onResize: onRatioChange,
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TreeMapLegendRenderer.TreeMapLegendRenderer, {})
          }
        )
      ]
    }
  );
});
