// @dynatrace/strato-components/editors/code-editor/CodeEditor.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CodeEditor_exports = {};
__export(CodeEditor_exports, {
  CodeEditor: () => CodeEditor,
  getLanguageConfig: () => getLanguageConfig
});
module.exports = __toCommonJS(CodeEditor_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_codeEditorMessages = require("./codeEditorMessages.js");
var import_languages = require("./languages/languages.js");
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useSafeForwardProps = require("../../core/hooks/useSafeForwardProps.js");
var import_BaseCodeEditor = require("../base-code-editor/BaseCodeEditor.js");
const COMPONENT_NAME = "CodeEditor";
const CodeEditor = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const intl = (0, import_react_intl.useIntl)();
  const {
    id,
    controlState,
    defaultFolding,
    folding,
    onFoldingChange,
    spellCheck = false,
    language = "other",
    placeholder = intl.formatMessage(import_codeEditorMessages.PLACEHOLDER_MESSAGES[language]),
    readOnly = false,
    lineWrap = false,
    fullHeight = false,
    value = "",
    onChange,
    onBlur,
    onFocus,
    gutterConfiguration,
    size = "default",
    required = false,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const ariaLabelingProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(restProps);
  const forwardProps = (0, import_useSafeForwardProps.useSafeForwardProps)(restProps, ariaLabelingProps);
  const config = (0, import_react.useMemo)(() => {
    return getLanguageConfig(import_languages.codeEditorLanguages[language], readOnly);
  }, [language, readOnly]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_BaseCodeEditor.BaseCodeEditor,
    {
      id,
      ref: forwardedRef,
      placeholder,
      controlState,
      readOnly,
      folding,
      defaultFolding,
      onFoldingChange,
      lineWrap,
      value,
      onChange,
      spellCheck,
      fullHeight,
      configuration: config,
      onBlur,
      onFocus,
      gutterConfiguration,
      size,
      required,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...ariaLabelingProps,
      ...behavioralTrackingProps,
      ...forwardProps,
      "data-dt-component": COMPONENT_NAME
    }
  );
});
function getLanguageConfig(languageConfig, readOnly) {
  return {
    language: languageConfig.language,
    autocompleteOptions: readOnly ? void 0 : {
      activateOnTyping: true,
      providers: languageConfig.autocompleteOptions?.providers ?? []
    },
    extendWith: languageConfig.extendWith
  };
}
CodeEditor.displayName = COMPONENT_NAME;
