// @dynatrace/strato-components/editors/code-editor/codeEditorMessages.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var codeEditorMessages_exports = {};
__export(codeEditorMessages_exports, {
  PLACEHOLDER_MESSAGES: () => PLACEHOLDER_MESSAGES
});
module.exports = __toCommonJS(codeEditorMessages_exports);
var import_react_intl = require("react-intl");
const PLACEHOLDER_MESSAGES = (0, import_react_intl.defineMessages)({
  other: {
    defaultMessage: "Add code",
    description: "Default placeholder for code editor when no language is set.",
    id: "fIr5C8WQPnYfAMr5"
  },
  json: {
    defaultMessage: "Add JSON",
    description: "Default placeholder for code editor when the language is JSON.",
    id: "XXsHBFWgU7wZSd8V"
  },
  js: {
    defaultMessage: "Add JavaScript",
    description: "Default placeholder for code editor when the language is JavaScript.",
    id: "fg8x33u8t5tYtlE9"
  },
  jsx: {
    defaultMessage: "Add JSX",
    description: "Default placeholder for code editor when the language is JSX.",
    id: "PCUD8a2q2mhgGcgg"
  },
  ts: {
    defaultMessage: "Add TypeScript",
    description: "Default placeholder for code editor when the language is TypeScript.",
    id: "xT9m8P49DpYVR1pg"
  },
  tsx: {
    defaultMessage: "Add TSX",
    description: "Default placeholder for code editor when the language is TSX.",
    id: "9AzZAjitLsfniU2y"
  },
  md: {
    defaultMessage: "Add Markdown",
    description: "Default placeholder for code editor when the language is Markdown.",
    id: "ZrxrsiqabWfxne/q"
  },
  yaml: {
    defaultMessage: "Add YAML",
    description: "Default placeholder for code editor when the language is YAML.",
    id: "0mIAvvhLoJG6DVyY"
  }
});
