// @dynatrace/strato-components/esm/charts/pie/providers/ChartProviders.js
import { jsx } from "react/jsx-runtime";
import { StoreProvider } from "./StoreProvider.js";
import { ChartSeriesActionContext } from "../../core/components/series-action/ChartSeriesActionContext.js";
import { ChartLegendContext } from "../context/ChartLegendContext.js";
import { FormatterContext } from "../context/FormatterContext.js";
import { LabelsContext } from "../context/LabelsContext.js";
import { PieChartConfigContext } from "../context/PieChartConfigContext.js";
import { PieChartRawDataContext } from "../context/PieChartRawDataContext.js";
import {
  PieChartTooltipContext,
  DEFAULT_TOOLTIP_CONFIG
} from "../context/PieChartTooltipContext.js";
import { ToolbarContext } from "../context/ToolbarContext.js";
const ChartProviders = (props) => {
  const { chartConfig, chartData, children } = props;
  return /* @__PURE__ */ jsx(
    ChartSeriesActionContext.Provider,
    {
      value: {
        // We need a cast here as the provider expects unknown and the
        // react context will not be able to carry the generic anyways.
        chartSeriesActions: chartConfig.seriesActions
      },
      children: /* @__PURE__ */ jsx(PieChartRawDataContext.Provider, { value: chartData, children: /* @__PURE__ */ jsx(PieChartConfigContext.Provider, { value: chartConfig, children: /* @__PURE__ */ jsx(
        PieChartTooltipContext.Provider,
        {
          value: chartConfig.tooltip ?? DEFAULT_TOOLTIP_CONFIG,
          children: /* @__PURE__ */ jsx(
            LabelsContext.Provider,
            {
              value: {
                hidden: chartConfig.labels?.hidden,
                valueType: chartConfig.labels?.valueType
              },
              children: /* @__PURE__ */ jsx(
                FormatterContext.Provider,
                {
                  value: {
                    formatter: chartConfig.formatter,
                    truncationMode: chartConfig.truncationMode
                  },
                  children: /* @__PURE__ */ jsx(ChartLegendContext.Provider, { value: chartConfig.legend, children: /* @__PURE__ */ jsx(ToolbarContext.Provider, { value: chartConfig.toolbar, children: /* @__PURE__ */ jsx(StoreProvider, { children }) }) })
                }
              )
            }
          )
        }
      ) }) })
    }
  );
};
export {
  ChartProviders
};
