// @dynatrace/strato-components/charts/categorical-bar/components/interactions/overlays/CategoricalBarChartInspectOverlay.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CategoricalBarChartInspectOverlay_exports = {};
__export(CategoricalBarChartInspectOverlay_exports, {
  CROSSHAIR_BAR_CHART_LABEL: () => CROSSHAIR_BAR_CHART_LABEL,
  CategoricalBarChartInspectOverlay: () => CategoricalBarChartInspectOverlay
});
module.exports = __toCommonJS(CategoricalBarChartInspectOverlay_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_CategoricalBarChartInspectOverlay_css = require("./CategoricalBarChartInspectOverlay.sty.js");
var import_ChartGrid_css = require("../../../../core/components/chart-grid/ChartGrid.sty.js");
var import_useGridPlotSize = require("../../../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_HorizontalCrosshair = require("../../../../core/components/crosshair/HorizontalCrosshair.js");
var import_get_magnifier_label = require("../../../../core/components/crosshair/utils/get-magnifier-label.js");
var import_VerticalCrosshair = require("../../../../core/components/crosshair/VerticalCrosshair.js");
var import_constants = require("../../../../core/components/threshold-bar/types/constants.js");
var import_useApplyCallbackOnScroll = require("../../../../core/hooks/useApplyCallbackOnScroll.js");
var import_selectors = require("../../../hooks/store/selectors.js");
var import_state_setters = require("../../../hooks/store/state-setters.js");
var import_useCategoricalBarChartBaseConfig = require("../../../hooks/useCategoricalBarChartBaseConfig.js");
var import_useCategoricalBarChartData = require("../../../hooks/useCategoricalBarChartData.js");
var import_useCategoricalBarChartOverlayRef = require("../../../hooks/useCategoricalBarChartOverlayRef.js");
var import_useCategoricalBarChartScales = require("../../../hooks/useCategoricalBarChartScales.js");
var import_useCategoricalBarChartThresholds = require("../../../hooks/useCategoricalBarChartThresholds.js");
var import_useClosestDimensionInThresholdPoints = require("../../../hooks/useClosestDimensionInThresholdPoints.js");
var import_useClosestDimensionPoint = require("../../../hooks/useClosestDimensionPoint.js");
var import_useFormatter = require("../../../hooks/useFormatter.js");
var import_layout = require("../../../utils/layout.js");
var import_CategoricalBarChartDatapoints = require("../datapoints/CategoricalBarChartDatapoints.js");
var import_CategoricalBarChartTooltip = require("../tooltip/CategoricalBarChartTooltip.js");
var import_data_points = require("../utils/data-points.js");
const CROSSHAIR_BAR_CHART_LABEL = "crosshair for bar chart";
const CategoricalBarChartInspectOverlay = () => {
  const { layout, groupMode } = (0, import_useCategoricalBarChartBaseConfig.useCategoricalBarChartBaseConfig)();
  const { width, height } = (0, import_useGridPlotSize.useGridPlotSize)();
  const { unit } = (0, import_useCategoricalBarChartData.useCategoricalBarChartData)();
  const formatter = (0, import_useFormatter.useFormatter)();
  const { valueScale } = (0, import_useCategoricalBarChartScales.useCategoricalBarChartScales)();
  const { visibleTooltip, pinnedTooltip } = (0, import_selectors.useTooltipState)();
  const mousePosition = (0, import_selectors.useMousePositionState)();
  const thresholds = (0, import_useCategoricalBarChartThresholds.useCategoricalBarChartThresholds)();
  const chartRef = (0, import_useCategoricalBarChartOverlayRef.useCategoricalBarChartOverlayRef)();
  const isInChartBounds = (0, import_selectors.useMouseInBoundsState)();
  const chartBounds = chartRef?.current?.getBoundingClientRect();
  const hasThresholds = thresholds && thresholds.length > 0;
  const [pinnedClosestPoint, setPinnedClosestPoint] = (0, import_react.useState)();
  const [pinnedGroupDataPoint, setPinnedGroupDataPoint] = (0, import_react.useState)();
  const magnifierCoordinate = (0, import_layout.isVertical)(layout) ? "y" : "x";
  const magnifierValue = mousePosition && (0, import_get_magnifier_label.getMagnifierLabel)(
    unit,
    magnifierCoordinate,
    formatter,
    mousePosition.relative,
    valueScale
  );
  const isGroupedAllowed = (0, import_layout.isStacked)(groupMode);
  const { closestDimensionPoint, closestDimensionPoints } = (0, import_useClosestDimensionPoint.useClosestDimensionPoint)(layout, groupMode, mousePosition?.relative);
  const closestDimensionPointInThreshold = (0, import_useClosestDimensionInThresholdPoints.useClosestDimensionInThresholdPoints)(
    layout,
    groupMode,
    mousePosition?.relative
  );
  const groupDataPoints = (0, import_data_points.generateGroupDataPoints)(
    closestDimensionPoints,
    isGroupedAllowed,
    layout,
    layout === "vertical" ? height : width
  );
  (0, import_react.useEffect)(() => {
    if (!visibleTooltip && !pinnedTooltip) {
      setPinnedClosestPoint(
        (prevState) => prevState !== void 0 ? void 0 : prevState
      );
      setPinnedGroupDataPoint(
        (prevState) => prevState !== void 0 ? void 0 : prevState
      );
    }
  }, [visibleTooltip, pinnedTooltip]);
  (0, import_react.useEffect)(() => {
    if (pinnedTooltip) {
      onClickHandler();
    }
  }, [width, height]);
  const setState = (0, import_selectors.useSetState)();
  const onScrollCallback = () => setState(import_state_setters.setMouseOutBounds);
  const notShowingMagnifiers = pinnedTooltip || !isInChartBounds;
  (0, import_useApplyCallbackOnScroll.useApplyCallbackOnScroll)(notShowingMagnifiers, onScrollCallback);
  const onClickHandler = () => {
    setState((prev) => ({
      ...prev,
      overlays: {
        ...prev.overlays,
        pinnedTooltip: setPinned(prev.overlays.pinnedTooltip)
      }
    }));
    setPinnedClosestPoint(
      (prev) => !prev && visibleTooltip ? closestDimensionPointInThreshold ?? closestDimensionPoint : void 0
    );
    setPinnedGroupDataPoint((prev) => {
      if (!prev && closestDimensionPoint && visibleTooltip) {
        const dimensionId = (closestDimensionPointInThreshold ?? closestDimensionPoint)?.datapoint.id;
        return (0, import_data_points.getGroupDataPointsBy)(groupDataPoints, dimensionId);
      }
      return void 0;
    });
  };
  const setPinned = (prevPinned) => visibleTooltip ? !prevPinned : false;
  const verticalCrosshairX = pinnedTooltip && pinnedClosestPoint ? pinnedClosestPoint.x ?? 0 : mousePosition?.relative.x ?? 0;
  const verticalCrosshairLabel = pinnedTooltip || !mousePosition ? "" : magnifierValue;
  const horizontalCrosshairY = pinnedTooltip && pinnedClosestPoint ? pinnedClosestPoint.y ?? 0 : mousePosition?.relative.y ?? 0;
  const horizontalCrosshairLabel = pinnedTooltip || !mousePosition ? "" : magnifierValue;
  const closestPoint = pinnedTooltip && pinnedClosestPoint ? pinnedClosestPoint : closestDimensionPointInThreshold;
  const groupPoints = pinnedGroupDataPoint ? pinnedGroupDataPoint : groupDataPoints;
  const thresholdSize = hasThresholds ? import_constants.THRESHOLD_BAR_THICKNESS : 0;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { className: import_ChartGrid_css.plotLayout, children: [
    (mousePosition || pinnedTooltip) && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: import_CategoricalBarChartInspectOverlay_css.dataPointLayout, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_CategoricalBarChartDatapoints.CategoricalBarChartDatapoints,
        {
          points: groupPoints,
          closestPoint: closestPoint && (0, import_data_points.mapToDataPoint)(closestPoint)
        }
      ) }),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_CategoricalBarChartTooltip.CategoricalBarChartTooltip,
        {
          closestDimensionPoint: closestPoint,
          hasHeader: groupPoints.length > 1,
          formatter,
          unit
        }
      ) })
    ] }),
    pinnedTooltip || isInChartBounds ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "svg",
      {
        className: import_CategoricalBarChartInspectOverlay_css.crosshairLayout,
        "aria-label": CROSSHAIR_BAR_CHART_LABEL,
        role: "graphics-symbol",
        onClick: onClickHandler,
        children: !(0, import_layout.isVertical)(layout) ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_VerticalCrosshair.VerticalCrosshair,
          {
            x: verticalCrosshairX,
            bottomLabel: verticalCrosshairLabel,
            height: height + thresholdSize,
            offset: thresholdSize,
            dashed: !pinnedTooltip,
            chartBounds
          }
        ) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_HorizontalCrosshair.HorizontalCrosshair,
          {
            width: width + thresholdSize,
            leftLabel: horizontalCrosshairLabel,
            leftOffset: thresholdSize,
            y: horizontalCrosshairY,
            dashed: !pinnedTooltip,
            chartBounds
          }
        )
      }
    ) : null
  ] });
};
CategoricalBarChartInspectOverlay["displayName"] = "CategoricalBarChartInspectOverlay";
