// @dynatrace/strato-components/esm/forms/select/styles/Select.sty.js
import "../../../core/components/app-root/AppRoot.css";
import "../../../styles/field.css";
import "./Select.css";
import { createRuntimeFn as _7a468 } from "@vanilla-extract/recipes/createRuntimeFn";
var clearReactSelectZIndex = "_dlenfc5-3-6-0";
var filter = "_dlenfc1-3-6-0";
var placeholderColor = "_dlenfcl-3-6-0";
var select = "_dlenfc0-3-6-0";
var selectAll = "_dlenfcs-3-6-0";
var selectGroupHeading = "_dlenfcf-3-6-0";
var selectMenuDivider = "_dlenfcn-3-6-0";
var selectMenuListWrapper = "_dlenfch-3-6-0";
var selectMenuLoadingWrapper = "_dlenfcm-3-6-0";
var selectMenuOverlay = _7a468({ defaultClassName: "_dlenfco-3-6-0", variantClassNames: { isTriggerWidth: { true: "_dlenfcp-3-6-0" } }, defaultVariants: { isTriggerWidth: false }, compoundVariants: [] });
var selectMultiValue = "_dlenfcq-3-6-0";
var selectNoOptionsMessage = "_dlenfci-3-6-0";
var selectOptionMulti = "_dlenfct-3-6-0";
var selectOptionMultiIndicator = _7a468({ defaultClassName: "_dlenfcu-3-6-0", variantClassNames: { selected: { true: "_dlenfcv-3-6-0" } }, defaultVariants: {}, compoundVariants: [] });
var selectOptionMultiIndicatorIcon = "_dlenfcw-3-6-0";
var selectOptionPrefix = "_dlenfcx-3-6-0";
var selectOptionSingle = "_dlenfcr-3-6-0";
var selectOptionSuffix = "_dlenfcy-3-6-0";
var selectTrigger = _7a468({ defaultClassName: "_dlenfc9-3-6-0", variantClassNames: { fullWidth: { true: "_dlenfca-3-6-0" }, showError: { true: "_dlenfcb-3-6-0" }, readOnly: { true: "_dlenfcc-3-6-0" } }, defaultVariants: { fullWidth: false }, compoundVariants: [[{ showError: true, readOnly: true }, "_dlenfcd-3-6-0"], [{ showError: false, readOnly: true }, "_dlenfce-3-6-0"]] });
var selectTriggerIconWrapper = _7a468({ defaultClassName: "_dlenfc6-3-6-0", variantClassNames: { isOverlayOpen: { true: "_dlenfc7-3-6-0" }, readOnly: { true: "_dlenfc8-3-6-0" } }, defaultVariants: { isOverlayOpen: false }, compoundVariants: [] });
var selectValueContainerTruncationBox = "_dlenfcj-3-6-0";
var selectVirtualizationWrapper = "_dlenfcg-3-6-0";
var selectWrapper = _7a468({ defaultClassName: "_dlenfc2-3-6-0", variantClassNames: { fullWidth: { true: "_dlenfc3-3-6-0", false: "_dlenfc4-3-6-0" } }, defaultVariants: { fullWidth: false }, compoundVariants: [] });
var validAndEnabled = "_dlenfck-3-6-0";
export {
  clearReactSelectZIndex,
  filter,
  placeholderColor,
  select,
  selectAll,
  selectGroupHeading,
  selectMenuDivider,
  selectMenuListWrapper,
  selectMenuLoadingWrapper,
  selectMenuOverlay,
  selectMultiValue,
  selectNoOptionsMessage,
  selectOptionMulti,
  selectOptionMultiIndicator,
  selectOptionMultiIndicatorIcon,
  selectOptionPrefix,
  selectOptionSingle,
  selectOptionSuffix,
  selectTrigger,
  selectTriggerIconWrapper,
  selectValueContainerTruncationBox,
  selectVirtualizationWrapper,
  selectWrapper,
  validAndEnabled
};
