// @dynatrace/strato-components/esm/charts/categorical-bar/hooks/store/selectors.js
import { isEqual } from "lodash-es";
import { useContext, useRef } from "react";
import { useStoreSelector } from "../../../../core/store/selector.js";
import { CategoricalBarChartStateContext } from "../../contexts/CategoricalBarChartState.context.js";
const useSelector = (selector, isEqual2) => useStoreSelector(CategoricalBarChartStateContext, selector, isEqual2);
const useHighlightedDimensionState = () => useSelector((state) => state.highlightedDimension);
const useFilteredDimensionsState = () => useSelector((state) => state.filteredDimensions);
const useTooltipState = () => useSelector(
  (state) => ({
    pinnedTooltip: state.overlays.pinnedTooltip,
    visibleTooltip: state.overlays.visibleTooltip
  }),
  isEqual
);
const useHoveredThresholdIdState = () => useSelector((state) => state.hoveredThresholdId);
const useSetState = () => {
  const { setState } = useContext(CategoricalBarChartStateContext);
  return setState;
};
const useMousePositionState = () => {
  const ref = useRef(void 0);
  return useSelector((state) => {
    if (!state.overlays.pinnedTooltip) {
      ref.current = state.overlays.mousePosition;
    }
    return ref.current;
  }, isEqual);
};
const useMouseInBoundsState = () => useSelector(
  (state) => state.overlays.mouseInBounds
);
export {
  useFilteredDimensionsState,
  useHighlightedDimensionState,
  useHoveredThresholdIdState,
  useMouseInBoundsState,
  useMousePositionState,
  useSelector,
  useSetState,
  useTooltipState
};
