// @dynatrace/strato-components/esm/filters/filter-field/suggestions/AriaLiveInfo.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { useMemo, useEffect, useState } from "react";
import { FormattedMessage } from "react-intl";
import { VisuallyHidden } from "../../../core/components/visually-hidden/VisuallyHidden.js";
import { canReceiveDefaultTextFormat as _canReceiveDefaultTextFormat } from "../../../core/utils/can-receive-default-text-format.js";
import { useFilterFieldContext } from "../hooks/useFilterFieldContext.js";
import { filterFieldMessages } from "../messages.js";
const MAX_ANNOUNCED_SUGGESTIONS = 5;
const ARIA_LIVE_DEBOUNCE_MS = 600;
const DefaultSuggestionsAriaLiveInfo = ({
  suggestions
}) => {
  if (!suggestions.length) {
    return null;
  }
  return /* @__PURE__ */ jsx("ul", { children: suggestions.slice(0, MAX_ANNOUNCED_SUGGESTIONS).map((suggestion, index) => {
    let suggestionText;
    if (_canReceiveDefaultTextFormat(suggestion.displayValue)) {
      suggestionText = suggestion.displayValue;
    } else if (_canReceiveDefaultTextFormat(suggestion.value)) {
      suggestionText = suggestion.value;
    }
    if (suggestionText) {
      return /* @__PURE__ */ jsx("li", { children: suggestionText }, `${suggestionText}-${index}`);
    }
    return null;
  }) });
};
const LogicalOperatorSuggestionAriaLiveInfo = ({
  hasLogicalOperators
}) => {
  if (!hasLogicalOperators) {
    return null;
  }
  return /* @__PURE__ */ jsx("li", { children: /* @__PURE__ */ jsx(
    FormattedMessage,
    {
      ...filterFieldMessages.suggestions.ariaLiveInfo.logicalOperator
    }
  ) });
};
const GeneralSuggestionsAriaLiveInfo = ({
  values
}) => values.map((value) => /* @__PURE__ */ jsx("li", { children: value.displayValue }, value.value));
const RecentAndPinnedSuggestionsAriaLiveInfo = ({
  message,
  count
}) => {
  if (count === 0) {
    return null;
  }
  return /* @__PURE__ */ jsx("li", { children: /* @__PURE__ */ jsx(
    FormattedMessage,
    {
      ...message,
      values: {
        value: count
      }
    }
  ) });
};
const extractSuggestions = (flattenedItems, limit = MAX_ANNOUNCED_SUGGESTIONS) => {
  const result = [];
  for (const item of flattenedItems) {
    if (result.length >= limit) {
      break;
    }
    if (item.kind === "suggestion") {
      result.push(item.data);
    }
  }
  return result;
};
const AriaLiveInfo = ({
  flattenedSuggestionItems
}) => {
  const {
    isValidSearch,
    pastedContentSuggestionValues,
    logicalOperatorSuggestionValues,
    recentSuggestionValues,
    pinnedSuggestionValues,
    searchSuggestionValues,
    parserConfig
  } = useFilterFieldContext();
  const showSearchSuggestions = isValidSearch?.current && searchSuggestionValues.length > 0 && // eslint-disable-next-line @typescript-eslint/no-deprecated
  (parserConfig?.searchConversion ?? false);
  const hasAdditionalContent = pastedContentSuggestionValues.length > 0 || logicalOperatorSuggestionValues.length > 0 || showSearchSuggestions || recentSuggestionValues.length > 0 || pinnedSuggestionValues.length > 0;
  const combinedSuggestions = useMemo(
    () => extractSuggestions(flattenedSuggestionItems, MAX_ANNOUNCED_SUGGESTIONS),
    [flattenedSuggestionItems]
  );
  const [debouncedState, setDebouncedState] = useState(
    () => ({
      combinedSuggestions: [],
      hasAdditionalContent: false,
      showSearchSuggestions: false,
      pastedContentSuggestionValues: [],
      logicalOperatorSuggestionsAvailable: false,
      recentSuggestionCount: 0,
      pinnedSuggestionCount: 0,
      searchSuggestionValues: []
    })
  );
  useEffect(() => {
    const handle = setTimeout(() => {
      setDebouncedState({
        combinedSuggestions,
        hasAdditionalContent,
        showSearchSuggestions: !!showSearchSuggestions,
        pastedContentSuggestionValues,
        logicalOperatorSuggestionsAvailable: logicalOperatorSuggestionValues.length > 0,
        recentSuggestionCount: recentSuggestionValues.length,
        pinnedSuggestionCount: pinnedSuggestionValues.length,
        searchSuggestionValues
      });
    }, ARIA_LIVE_DEBOUNCE_MS);
    return () => {
      clearTimeout(handle);
    };
  }, [
    combinedSuggestions,
    hasAdditionalContent,
    showSearchSuggestions,
    pastedContentSuggestionValues,
    logicalOperatorSuggestionValues.length,
    recentSuggestionValues.length,
    pinnedSuggestionValues.length,
    searchSuggestionValues
  ]);
  const hasContent = debouncedState.combinedSuggestions.length > 0;
  return /* @__PURE__ */ jsx(VisuallyHidden, { children: /* @__PURE__ */ jsx("div", { "aria-live": "polite", "aria-atomic": "true", children: hasContent || debouncedState.hasAdditionalContent ? /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(
      FormattedMessage,
      {
        ...filterFieldMessages.suggestions.ariaLiveInfo.availableSuggestions
      }
    ),
    /* @__PURE__ */ jsx(
      DefaultSuggestionsAriaLiveInfo,
      {
        suggestions: debouncedState.combinedSuggestions
      }
    ),
    debouncedState.hasAdditionalContent && /* @__PURE__ */ jsxs("ul", { children: [
      /* @__PURE__ */ jsx(
        GeneralSuggestionsAriaLiveInfo,
        {
          values: debouncedState.pastedContentSuggestionValues
        }
      ),
      /* @__PURE__ */ jsx(
        LogicalOperatorSuggestionAriaLiveInfo,
        {
          hasLogicalOperators: debouncedState.logicalOperatorSuggestionsAvailable
        }
      ),
      /* @__PURE__ */ jsx(
        RecentAndPinnedSuggestionsAriaLiveInfo,
        {
          message: filterFieldMessages.suggestions.ariaLiveInfo.recentlyUsedSuggestions,
          count: debouncedState.recentSuggestionCount
        }
      ),
      /* @__PURE__ */ jsx(
        RecentAndPinnedSuggestionsAriaLiveInfo,
        {
          message: filterFieldMessages.suggestions.ariaLiveInfo.pinnedSuggestions,
          count: debouncedState.pinnedSuggestionCount
        }
      ),
      /* @__PURE__ */ jsx(
        GeneralSuggestionsAriaLiveInfo,
        {
          values: debouncedState.searchSuggestionValues
        }
      )
    ] }),
    /* @__PURE__ */ jsx(
      FormattedMessage,
      {
        ...filterFieldMessages.suggestions.ariaLiveInfo.moreSuggestionsAvailable
      }
    )
  ] }) : null }) });
};
export {
  AriaLiveInfo
};
