// @dynatrace/strato-components/esm/filters/filter-field/suggestions/FallbackKeySuggestions.js
import { jsx, jsxs } from "react/jsx-runtime";
import { createElement } from "react";
import { useIntl } from "react-intl";
import { CSSHighlight } from "../components/CSSHighlight.js";
import { useFilterFieldContext } from "../hooks/useFilterFieldContext.js";
import { filterFieldMessages } from "../messages.js";
import * as suggestionStyles from "./Suggestion.sty.js";
import { Suggestion } from "./Suggestion.js";
import * as suggestionGroupStyles from "./SuggestionGroup.sty.js";
import { SuggestionGroup } from "./SuggestionGroup.js";
function FallbackKeySuggestionGroup() {
  const { fallbackKeySuggestionValues, statementSuggestionValues, term } = useFilterFieldContext();
  const intl = useIntl();
  if (fallbackKeySuggestionValues.length === 0 && statementSuggestionValues.length === 0) {
    return null;
  }
  return /* @__PURE__ */ jsxs(
    SuggestionGroup,
    {
      "aria-label": intl.formatMessage(
        filterFieldMessages.suggestions.a11y.statementSuggestions
      ),
      children: [
        /* @__PURE__ */ jsx(
          CSSHighlight,
          {
            className: suggestionGroupStyles.builtInSuggestionsContainer,
            term: term ?? "",
            selector: `.${suggestionStyles.statementSuggestionValue}`,
            highlightName: "filterfield-suggestions-highlight",
            children: statementSuggestionValues?.map((suggestion, index) => /* @__PURE__ */ createElement(
              Suggestion,
              {
                ...suggestion,
                key: suggestion.value ?? `${index}-statementsuggestions`,
                "data-strato-filter-field-virtual-focus-type": "statement-suggestion"
              }
            ))
          }
        ),
        fallbackKeySuggestionValues?.map((suggestion, index) => /* @__PURE__ */ createElement(
          Suggestion,
          {
            ...suggestion,
            key: suggestion.value ?? `${index}-fallbackkeysuggestions`,
            className: suggestionStyles.defaultSuggestion,
            "data-strato-filter-field-virtual-focus-type": "fallbackkey-suggestion"
          }
        ))
      ]
    }
  );
}
export {
  FallbackKeySuggestionGroup
};
