// @dynatrace/strato-components/esm/filters/filter-field/suggestions/RecentPinnedSuggestionItem.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { useEffect, useMemo, useRef } from "react";
import { FormattedMessage } from "react-intl";
import { Button } from "../../../buttons/button/Button.js";
import { useId } from "../../../core/hooks/useId.js";
import { getOrCreateNamedHighlight } from "../components/CSSHighlight.js";
import { FilterStatement } from "../components/FilterStatement.js";
import { useFilterFieldContext } from "../hooks/useFilterFieldContext.js";
import { useFilterFieldLoggerContext } from "../hooks/useFilterFieldLoggerContext.js";
import { useShowMore } from "../hooks/useShowMore.js";
import { filterFieldMessages } from "../messages.js";
import * as styles from "./RecentPinnedSuggestionItem.sty.js";
import * as suggestionStyles from "./Suggestion.sty.js";
import { Suggestion } from "./Suggestion.js";
import { SuggestionGroup } from "./SuggestionGroup.js";
import { SuggestionGroupLabel } from "./SuggestionGroupLabel.js";
import { SuggestionPin } from "./SuggestionPin.js";
function RecentAndPinnedSuggestionItem(props) {
  const { suggestion, index, ranges, ...remainingProps } = props;
  const suggestionRef = useRef(null);
  const id = useId();
  const { logger } = useFilterFieldLoggerContext();
  const suggestionHighlight = useMemo(
    () => getOrCreateNamedHighlight("filterfield-suggestions-highlight"),
    []
  );
  useEffect(() => {
    const highlightRanges = [];
    if (suggestionRef.current) {
      const valueElement = suggestionRef.current.querySelector(
        `.${suggestionStyles.value} span`
      );
      if (valueElement && valueElement.textContent !== "") {
        const textNode = valueElement.childNodes.item(0);
        if (textNode.nodeType === Node.TEXT_NODE) {
          const textLength = textNode.textContent?.length ?? 0;
          for (const [start, end] of ranges) {
            const clampedStart = Math.min(start, textLength);
            const clampedEnd = Math.min(end, textLength);
            if (clampedStart >= clampedEnd) {
              continue;
            }
            try {
              const range = new Range();
              range.setStart(textNode, clampedStart);
              range.setEnd(textNode, clampedEnd);
              suggestionHighlight.add(range);
              highlightRanges.push(range);
            } catch (error) {
              logger?.warn(
                "FilterField suggestion highlight range error",
                error
              );
            }
          }
        }
      }
    }
    return () => {
      if (highlightRanges) {
        for (const range of highlightRanges) {
          suggestionHighlight.delete(range);
        }
      }
    };
  }, [logger, ranges, suggestionHighlight]);
  return /* @__PURE__ */ jsxs(
    Suggestion,
    {
      className: styles.recentSuggestion,
      ref: suggestionRef,
      ...suggestion,
      ...remainingProps,
      children: [
        /* @__PURE__ */ jsx(
          FilterStatement,
          {
            value: suggestion.displayValue?.toString(),
            highlighting: "minimal"
          }
        ),
        suggestion.value ? /* @__PURE__ */ jsx(
          SuggestionPin,
          {
            role: "option",
            filterString: suggestion.value,
            className: styles.suggestionPin,
            placement: "right",
            fallbackPlacements: ["right", "left"],
            id,
            "data-strato-filter-field-virtual-focus-id": id,
            "data-strato-filter-field-virtual-focus-type": "pinned-suggestion"
          }
        ) : null
      ]
    },
    `${suggestion.value ?? "non-interactive-"}${index}`
  );
}
function RecentSuggestionGroup() {
  const { filterFieldId, recentSuggestionValues } = useFilterFieldContext();
  let primaryCount = 2;
  if (recentSuggestionValues.length === 3) {
    primaryCount = 3;
  }
  const {
    items: slicedRecentSuggestionValues,
    isSliced,
    controlProps
  } = useShowMore({
    items: recentSuggestionValues,
    primaryCount
  });
  if (recentSuggestionValues.length === 0) {
    return null;
  }
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsxs(
      SuggestionGroup,
      {
        "aria-labelledby": `${filterFieldId ?? ""}-recent-filters`,
        children: [
          /* @__PURE__ */ jsx(SuggestionGroupLabel, { id: `${filterFieldId ?? ""}-recent-filters`, children: /* @__PURE__ */ jsx(FormattedMessage, { ...filterFieldMessages.suggestions.recentlyUsed }) }),
          slicedRecentSuggestionValues?.map((recentItem, index) => /* @__PURE__ */ jsx(
            RecentAndPinnedSuggestionItem,
            {
              suggestion: recentItem.suggestion,
              index,
              "data-strato-filter-field-virtual-focus-type": "recent-suggestion",
              ranges: recentItem.ranges
            },
            recentItem.suggestion.value ?? index
          ))
        ]
      }
    ),
    isSliced && /* @__PURE__ */ jsx("div", { role: "presentation", children: /* @__PURE__ */ jsx(Button, { ...controlProps }) })
  ] });
}
function PinnedSuggestionGroup() {
  const { filterFieldId, pinnedSuggestionValues } = useFilterFieldContext();
  if (pinnedSuggestionValues.length === 0) {
    return null;
  }
  return /* @__PURE__ */ jsxs(SuggestionGroup, { "aria-labelledby": `${filterFieldId ?? ""}-pinned-filters`, children: [
    /* @__PURE__ */ jsx(SuggestionGroupLabel, { id: `${filterFieldId ?? ""}-pinned-filters`, children: /* @__PURE__ */ jsx(FormattedMessage, { ...filterFieldMessages.suggestions.pinnedFilters }) }),
    pinnedSuggestionValues?.map((pinnedItem, index) => /* @__PURE__ */ jsx(
      RecentAndPinnedSuggestionItem,
      {
        suggestion: pinnedItem.suggestion,
        index,
        "data-strato-filter-field-virtual-focus-type": "pinned-suggestion",
        ranges: pinnedItem.ranges
      },
      pinnedItem.suggestion.value ?? index
    ))
  ] });
}
export {
  PinnedSuggestionGroup,
  RecentSuggestionGroup
};
