// @dynatrace/strato-components/esm/filters/filter-field/suggestions/SuggestionDetails.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import * as styles from "./SuggestionDetails.sty.js";
import { TextEllipsisAroundChildren } from "../../../core/components/TextEllipsisAroundChildren.js";
import { useBehavioralTrackingProps } from "../../../core/hooks/useBehavioralTrackingProps.js";
import { canReceiveDefaultTextFormat as _canReceiveDefaultTextFormat } from "../../../core/utils/can-receive-default-text-format.js";
import { Flex } from "../../../layouts/flex/Flex.js";
import { Text } from "../../../typography/text/Text.js";
const SuggestionDetails = (props) => {
  const {
    children,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps] = useBehavioralTrackingProps(remainingProps);
  const childIsText = _canReceiveDefaultTextFormat(children);
  return /* @__PURE__ */ jsx(
    Flex,
    {
      className: clsx(styles.suggestionDetails, consumerClassName),
      style: consumerStyle,
      "data-testid": dataTestId,
      ...behavioralTrackingProps,
      alignItems: "center",
      paddingY: 2,
      children: childIsText ? /* @__PURE__ */ jsx(Text, { textStyle: "small", className: styles.suggestionDetailsText, children: /* @__PURE__ */ jsx(TextEllipsisAroundChildren, { children }) }) : /* @__PURE__ */ jsx(Text, { as: "div", textStyle: "small", children })
    }
  );
};
SuggestionDetails.displayValue = "FilterField.SuggestionDetails";
export {
  SuggestionDetails
};
