// @dynatrace/strato-components/esm/filters/filter-field/suggestions/suggestions-extension.js
import {
  autocompletion,
  startCompletion
} from "@codemirror/autocomplete";
import { syntaxTree } from "@codemirror/language";
import { EditorView } from "@codemirror/view";
import {
  ANNOTATION_DOC_CHANGED_CONTROLLED,
  ANNOTATION_SUGGESTION_APPLY,
  ANNOTATION_SUGGESTION_PIN,
  pastingMode
} from "../editor-state-configs/editor-config.js";
import { escapeSuggestion } from "../utils/escape-suggestion.js";
import { getCurrentToken } from "../utils/suggestions-utils/get-current-token.js";
import { getSuggestionCategory } from "../utils/suggestions-utils/get-suggestion-category.js";
import { getSuggestions } from "../utils/suggestions-utils/get-suggestions.js";
import {
  getFallbackKeySuggestions,
  getFilterStatementSuggestions,
  getSearchSuggestions
} from "../utils/suggestions-utils/key-suggestions.js";
import {
  getRecentAndPinnedSuggestions
} from "../utils/suggestions-utils/recent-and-pinned-suggestions.js";
import { getFilterStatement } from "../utils/suggestions-utils/suggestions-callback.js";
import { isCursorInsideGroup } from "../utils/tree-navigation.js";
const suggestionsExtension = ({
  overlayState,
  numberParser,
  internalValidatorMap,
  onSuggest,
  _onSuggest,
  getSyntaxTree,
  customTypes,
  parserConfig
}) => {
  const filterFieldSuggestion = (context) => {
    if (context.view?.composing) {
      return null;
    }
    const tree = syntaxTree(context.state);
    const { tree: convertedTree } = getSyntaxTree({
      tree,
      numberParser,
      doc: context.state.doc.toString(),
      customTypes,
      parserConfig
    });
    const pastingStateField = context.view?.state.field(pastingMode, false);
    const { head: cursorPosition } = context.state.selection.main;
    let currentStatement = getFilterStatement(
      convertedTree,
      cursorPosition,
      context.state.doc.toString()
    );
    const token = getCurrentToken({
      tree,
      convertedTree,
      cursorPosition,
      doc: context.state.doc.toString()
    });
    let currentToken = token?.currentToken;
    const currentTokenType = token?.currentTokenType;
    const treeCursor = tree.cursorAt(cursorPosition, -1);
    treeCursor.childBefore(cursorPosition);
    const prevNode = treeCursor.node.type.is("FilterValue") || treeCursor.node.parent?.type.is("FilterValue") ? "FilterValue" : treeCursor.node.name;
    const doc = context.state.doc;
    const {
      showInclusionOperators,
      showLogicalOperators,
      suggestionType,
      overlappingSuggestions
    } = getSuggestionCategory({
      currentStatement,
      currentToken,
      currentTokenType,
      prevNode,
      tree,
      doc: context.state.doc.toString(),
      cursorPosition
    });
    const suggestionTypes = Array.from(
      new Set(
        [suggestionType].concat(
          // Using concat with the ternary so that typescript can correctly infer the array's type.
          overlappingSuggestions ? overlappingSuggestions.type : []
        )
      )
    );
    let suggestionTypeForToken = suggestionTypes.map((type) => ({
      type,
      token: type === overlappingSuggestions?.type ? overlappingSuggestions.subToken : currentToken
    }));
    if (pastingStateField && pastingStateField.pastingInProgress && pastingStateField.range) {
      suggestionTypes.push("pastedContent");
      currentStatement = void 0;
      const textValue = doc.sliceString(
        pastingStateField.range.from,
        pastingStateField.range.to
      );
      const escapedTextValue = escapeSuggestion(textValue);
      const pastedContentToken = {
        range: {
          from: pastingStateField.range.from,
          to: pastingStateField.range.to
        },
        value: textValue,
        textValue: escapedTextValue,
        type: "String"
      };
      const { token: token2, type } = getCurrentTokenForPastedContent(
        currentToken,
        pastingStateField.range,
        context.state,
        pastedContentToken
      );
      if (type) {
        suggestionTypeForToken = [
          {
            type,
            token: token2
          }
        ];
        currentToken = token2;
      }
      suggestionTypeForToken.push({
        type: "pastedContent",
        token: pastedContentToken
      });
    }
    const {
      logicalOperatorSuggestions,
      groupedSuggestions,
      pastedContentSuggestions
    } = getSuggestions({
      suggestionTypes,
      overlappingSuggestions,
      currentStatement,
      currentToken,
      currentTokenType,
      prevNode,
      internalValidatorMap,
      customTypes,
      showLogicalOperators,
      showInclusionOperators,
      doc: context.state.doc.toString(),
      suggestionTypeForToken
    });
    const { recentSuggestionValues, pinnedSuggestionValues } = getRecentAndPinnedSuggestions(context);
    const searchSuggestionValues = getSearchSuggestions(
      context,
      currentStatement,
      currentToken,
      suggestionTypes,
      isCursorInsideGroup(context.state, cursorPosition)
    );
    let filterStatementSuggestions = [];
    if (currentToken.type !== "List") {
      filterStatementSuggestions = getFilterStatementSuggestions(
        currentToken,
        suggestionTypes,
        internalValidatorMap?.filterStatementKeysMap
      );
    }
    const fallbackKeySuggestions = getFallbackKeySuggestions(
      currentToken,
      suggestionTypes,
      internalValidatorMap?.fallbackKeys,
      filterStatementSuggestions
    );
    _onSuggest({
      suggestionsContext: {
        token: currentToken,
        currentStatement
      },
      autoSuggestionValues: groupedSuggestions,
      recentSuggestionValues,
      pinnedSuggestionValues,
      logicalOperatorSuggestionValues: logicalOperatorSuggestions,
      pastedContentSuggestionValues: pastedContentSuggestions,
      searchSuggestionValues,
      fallbackKeySuggestionValues: fallbackKeySuggestions,
      filterStatementSuggestionValues: filterStatementSuggestions
    });
    onSuggest?.({
      suggestionTypes,
      currentTokens: suggestionTypeForToken,
      value: context.state.doc.toString(),
      groupedSuggestions,
      syntaxTree: convertedTree,
      cursorPosition,
      currentStatement
    });
    overlayState.open();
    return null;
  };
  return [
    autocompletion({ override: [filterFieldSuggestion] }),
    EditorView.updateListener.of((update) => {
      if (update.view.composing) {
        return;
      }
      if (update.focusChanged && update.view.hasFocus || update.transactions.some(
        (transaction) => transaction.isUserEvent("input") || transaction.isUserEvent(ANNOTATION_SUGGESTION_APPLY) || transaction.isUserEvent(ANNOTATION_SUGGESTION_PIN) || transaction.isUserEvent("delete") || transaction.selection !== void 0 && (transaction.isUserEvent(ANNOTATION_DOC_CHANGED_CONTROLLED) || update.view.hasFocus)
      )) {
        startCompletion(update.view);
        return;
      }
      if (!update.view.hasFocus) {
        overlayState.close();
        return;
      }
    })
  ];
};
function getCurrentTokenForPastedContent(currentToken, pastingRange, state, pastedContentToken) {
  if (pastingRange.from >= currentToken.range.from && pastingRange.to <= currentToken.range.to) {
    return { token: currentToken, type: void 0 };
  }
  return getTokenBeforePastedContent(
    pastingRange.from,
    state,
    pastedContentToken
  );
}
function getTrailingNodeEnd(tree, position) {
  const trailingCursor = tree.cursorAt(position, 1);
  if ((trailingCursor.node.name === "FreeText" || trailingCursor.node.name === "Identifier") && trailingCursor.node.from <= position) {
    return trailingCursor.node.to;
  }
  return null;
}
function getTokenBeforePastedContent(pastedRangeFrom, editorState, pastedContentToken) {
  const tree = syntaxTree(editorState);
  const cursor = tree.cursorAt(pastedRangeFrom, -1);
  if (cursor.node.name === "Filter") {
    return { token: pastedContentToken, type: "key" };
  }
  if (cursor.node.name === "FilterStatement" || cursor.node.name === "ComparisonOperator") {
    const rangeTo = getTrailingNodeEnd(tree, pastedContentToken.range.to);
    if (rangeTo !== null) {
      const textValue = editorState.doc.sliceString(
        pastedContentToken.range.from,
        rangeTo
      );
      return {
        token: {
          range: { from: pastedContentToken.range.from, to: rangeTo },
          value: textValue,
          textValue: escapeSuggestion(textValue),
          type: "String"
        },
        type: "value"
      };
    }
    return { token: pastedContentToken, type: "value" };
  }
  if (cursor.node.name === "FreeText" || cursor.node.name === "Identifier") {
    const rangeFrom = cursor.node.from;
    const rangeTo = getTrailingNodeEnd(tree, pastedContentToken.range.to) ?? pastedContentToken.range.to;
    const textValue = editorState.doc.sliceString(rangeFrom, rangeTo);
    const combinedToken = {
      range: { from: rangeFrom, to: rangeTo },
      value: textValue,
      textValue: escapeSuggestion(textValue),
      type: "String"
    };
    return {
      token: combinedToken,
      type: cursor.node.name === "FreeText" ? "key" : "value"
    };
  }
  return { token: pastedContentToken, type: void 0 };
}
export {
  getTokenBeforePastedContent,
  suggestionsExtension
};
