// @dynatrace/strato-components/esm/filters/filter-field/suggestions/SuggestionsOverlay.js
import { jsx, jsxs } from "react/jsx-runtime";
import { motion } from "framer-motion";
import { forwardRef } from "react";
import * as styles from "./SuggestionsOverlay.sty.js";
import { SuggestionTips } from "./SuggestionTips.js";
import { OverlayContainer } from "../../../core/components/OverlayContainer.js";
import { useOverlayStacking } from "../../../core/hooks/useOverlayStacking.js";
import { OverlayStackingProvider } from "../../../core/providers/OverlayStackingProvider.js";
import useOverlayTransition from "../hooks/useOverlayTransition.js";
const suggestionsOverlayAnimationVariants = {
  inital: {
    opacity: 0.4,
    scaleY: 0.8
  },
  default: {
    opacity: 1,
    scaleY: 1
  },
  exit: {
    opacity: 0,
    scaleY: 0.8
  }
};
const SuggestionsOverlay = forwardRef((props, forwardedRef) => {
  const {
    children,
    overlayProps,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow
  } = props;
  const overlayContainerProps = useOverlayStacking(void 0, "dropdown");
  const transition = useOverlayTransition(false);
  return /* @__PURE__ */ jsx(
    OverlayContainer,
    {
      className: "suggestions-overlay",
      ...overlayContainerProps,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      children: /* @__PURE__ */ jsx(OverlayStackingProvider, { children: /* @__PURE__ */ jsx(
        "div",
        {
          ref: forwardedRef,
          "data-testid": dataTestId,
          onPointerDown: (event) => {
            event.preventDefault();
          },
          ...overlayProps,
          children: /* @__PURE__ */ jsxs(
            motion.div,
            {
              className: styles.suggestionsOverlayWrapper,
              variants: suggestionsOverlayAnimationVariants,
              initial: "inital",
              animate: "default",
              exit: "exit",
              transition,
              children: [
                children,
                /* @__PURE__ */ jsx(SuggestionTips, {})
              ]
            }
          )
        }
      ) })
    }
  );
});
export {
  SuggestionsOverlay
};
