// @dynatrace/strato-components/esm/filters/filter-field/suggestions/Suggestion.js
import { jsx, jsxs } from "react/jsx-runtime";
import { createElement } from "react";
import clsx from "clsx";
import { isArray } from "lodash-es";
import {
  cloneElement,
  forwardRef,
  useEffect,
  useMemo,
  useRef,
  useState
} from "react";
import { FormattedMessage } from "react-intl";
import * as styles from "./Suggestion.sty.js";
import { SuggestionDetails } from "./SuggestionDetails.js";
import { SuggestionPin } from "./SuggestionPin.js";
import { Chip } from "../../../content/chip/Chip.js";
import { useId } from "../../../core/hooks/useId.js";
import { useMergeRefs } from "../../../core/hooks/useMergeRefs.js";
import { useSlots } from "../../../core/hooks/useSlots.js";
import { getVirtualFocusStyles } from "../../../core/styles/getVirtualFocusStyles.js";
import { canReceiveDefaultTextFormat as _canReceiveDefaultTextFormat } from "../../../core/utils/can-receive-default-text-format.js";
import { Flex } from "../../../layouts/flex/Flex.js";
import { Grid } from "../../../layouts/grid/Grid.js";
import { Tooltip } from "../../../overlays/tooltip/Tooltip.js";
import {
  ANNOTATION_END_PASTING_MODE,
  pastingEffect,
  pastingMode,
  stringAnnotation
} from "../editor-state-configs/editor-config.js";
import { useFilterFieldContext } from "../hooks/useFilterFieldContext.js";
import { filterFieldMessages } from "../messages.js";
import { FilterFieldActions } from "../state/filter-field-state-reducer.js";
function WrappedSuggestionLabel({
  displayLabel
}) {
  if (typeof displayLabel === "string" || isArray(displayLabel)) {
    return /* @__PURE__ */ jsx("span", { className: styles.suggestionText, children: displayLabel });
  }
  return displayLabel;
}
function getDescription(type) {
  switch (type) {
    case "Boolean":
      return /* @__PURE__ */ jsx(
        FormattedMessage,
        {
          ...filterFieldMessages.suggestions.supportedValues.Boolean
        }
      );
    case "Duration":
      return /* @__PURE__ */ jsx(
        FormattedMessage,
        {
          ...filterFieldMessages.suggestions.supportedValues.Durations
        }
      );
    case "Number":
      return /* @__PURE__ */ jsx(
        FormattedMessage,
        {
          ...filterFieldMessages.suggestions.supportedValues.Numbers
        }
      );
    case "IPAddress":
      return /* @__PURE__ */ jsx(
        FormattedMessage,
        {
          ...filterFieldMessages.suggestions.supportedValues.IPAddresses
        }
      );
    case "UID":
      return /* @__PURE__ */ jsx(
        FormattedMessage,
        {
          ...filterFieldMessages.suggestions.supportedValues.UIDs
        }
      );
    case "SmartscapeId":
      return /* @__PURE__ */ jsx(
        FormattedMessage,
        {
          ...filterFieldMessages.suggestions.supportedValues.SmartscapeIds
        }
      );
    case "Timestamp":
      return /* @__PURE__ */ jsx(
        FormattedMessage,
        {
          ...filterFieldMessages.suggestions.supportedValues.Timestamps
        }
      );
    case "String":
    case "Any":
    default:
      return void 0;
  }
}
function dispatchEndPastingModeEffect(editorView) {
  editorView?.dispatch({
    effects: [
      pastingEffect.of({
        pastingInProgress: false
      })
    ],
    // In order to re-run the decorations and re-generate errors and diagnostics,
    // we have to dispatch a new document.
    changes: {
      from: 0,
      to: editorView?.state.doc.length,
      insert: editorView?.state.doc
    },
    selection: editorView.state.selection,
    annotations: stringAnnotation.of(ANNOTATION_END_PASTING_MODE)
  });
}
function handleSuggestionApply(editorView, id, applySuggestion, suggestionRef, suggestionText, trailing, insertionStrategy, insertionRange, cursorBeforeTrailing, escapeSuggestion) {
  if (id === "strato-filterfield-suggestion-pasted-content") {
    dispatchEndPastingModeEffect(editorView);
  } else {
    const pastingInProgress = editorView?.state.field(pastingMode, false);
    if (pastingInProgress?.pastingInProgress) {
      dispatchEndPastingModeEffect(editorView);
    }
    applySuggestion(
      suggestionRef,
      suggestionText,
      trailing,
      insertionStrategy,
      cursorBeforeTrailing,
      insertionRange,
      escapeSuggestion
    );
  }
}
const Suggestion = forwardRef((props, forwardedRef) => {
  const {
    value,
    displayValue,
    details,
    trailing = " ",
    cursorBeforeTrailing = false,
    insertionStrategy = "replace-token",
    insertionRange,
    escapeSuggestion,
    valueType,
    className: consumerClassName,
    style: consumerStyle,
    icon,
    "data-testid": dataTestId,
    "aria-describedby": ariaDescribedBy,
    "aria-labelledby": ariaLabelledBy,
    children,
    id,
    suggestionType,
    ...remainingProps
  } = props;
  const suggestionRef = useRef(null);
  const rootRef = useMergeRefs([forwardedRef, suggestionRef]);
  const { applySuggestion, dispatch, focusedListItem, editorView } = useFilterFieldContext();
  const valueRef = useRef(null);
  const generatedId = useId();
  const uniqueId = id ?? generatedId;
  const [slots, label] = useSlots(children, {
    details: SuggestionDetails,
    pin: SuggestionPin
  });
  const childIsText = _canReceiveDefaultTextFormat(label);
  const displayValueIsText = _canReceiveDefaultTextFormat(displayValue);
  const isFocused = uniqueId === focusedListItem?.id ? true : void 0;
  const finalDetails = useMemo(() => {
    if (slots.details) {
      return cloneElement(slots.details, {
        className: clsx(
          slots.details.props.className,
          styles.matchableSuggestionDetails
        )
      });
    }
    if (typeof details === "string") {
      return /* @__PURE__ */ jsx(SuggestionDetails, { className: styles.matchableSuggestionDetails, children: details });
    }
    if (details) {
      return /* @__PURE__ */ jsx(SuggestionDetails, { children: details });
    }
    if (suggestionType !== "validatorMap") {
      const description = getDescription(valueType);
      if (description) {
        return /* @__PURE__ */ jsx(SuggestionDetails, { children: description });
      }
    }
    return null;
  }, [details, slots.details, suggestionType, valueType]);
  const [showTooltip, setShowTooltip] = useState(false);
  if (value === void 0 && !childIsText && !displayValueIsText) {
    console.warn(
      "Suggestion value: There is no value provided for a suggestion and the children are not text-only. The applied result may be unexpected."
    );
  }
  let suggestionText = "";
  if (value) {
    suggestionText = value;
  } else if (childIsText) {
    suggestionText = label.join("");
  } else if (displayValueIsText) {
    suggestionText = displayValue;
  }
  if (typeof suggestionText === "string" && !trailing) {
    suggestionText = suggestionText.trim();
  }
  useEffect(() => {
    if (typeof suggestionText === "string") {
      dispatch({
        type: FilterFieldActions.ADDSUGGESTION,
        payload: {
          id: uniqueId,
          applySuggestion: () => {
            handleSuggestionApply(
              editorView,
              uniqueId,
              applySuggestion,
              suggestionRef.current,
              suggestionText,
              trailing,
              insertionStrategy,
              insertionRange,
              cursorBeforeTrailing,
              escapeSuggestion
            );
          }
        }
      });
    }
    return () => dispatch({
      type: FilterFieldActions.REMOVESUGGESTION,
      payload: { id: uniqueId }
    });
  }, []);
  if (typeof suggestionText !== "string") {
    return null;
  }
  const displayLabel = label.length > 0 ? label : displayValue ?? value;
  if (!value && trailing === "" && id !== "strato-filterfield-suggestion-pasted-content") {
    return /* @__PURE__ */ createElement(
      Flex,
      {
        ...remainingProps,
        role: "presentation",
        key: value,
        justifyContent: "space-between",
        className: id === "strato-filterfield-suggestion-jsonpath-hint" ? styles.suggestionsJSONPathHint : styles.suggestionsEmptyState,
        ref: rootRef
      },
      /* @__PURE__ */ jsx("span", { children: displayLabel }),
      finalDetails
    );
  }
  return /* @__PURE__ */ jsxs(
    Flex,
    {
      ...remainingProps,
      role: slots.pin && "group",
      gap: 0,
      "data-testid": dataTestId,
      className: clsx(styles.suggestion, consumerClassName),
      onTouchEnd: (e) => {
        e.preventDefault();
      },
      style: consumerStyle,
      ref: rootRef,
      children: [
        /* @__PURE__ */ jsx(
          Tooltip,
          {
            text: displayLabel,
            placement: "left",
            open: showTooltip,
            className: styles.tooltipWrapper,
            onOpenChange: (isOpen) => {
              const valueNodeChild = valueRef.current?.children[0];
              const hasOverFlow = (valueNodeChild?.scrollWidth ?? 0) > (valueNodeChild?.clientWidth ?? 0);
              setShowTooltip(isOpen && hasOverFlow);
            },
            children: /* @__PURE__ */ jsxs(
              Grid,
              {
                id: uniqueId,
                role: "option",
                "aria-selected": isFocused,
                "aria-labelledby": ariaLabelledBy,
                "aria-describedby": ariaDescribedBy,
                "data-strato-filter-field-virtual-focus-id": uniqueId,
                "data-strato-filter-field-virtual-focus-type": "suggestion",
                columnGap: 8,
                rowGap: 0,
                className: clsx(
                  styles.suggestionContent,
                  getVirtualFocusStyles(!!isFocused).className,
                  id === "strato-filterfield-suggestion-pasted-content" ? styles.pastingSuggestion : {}
                ),
                paddingLeft: icon ? 8 : 12,
                onPointerDown: () => {
                  handleSuggestionApply(
                    editorView,
                    id,
                    applySuggestion,
                    suggestionRef.current,
                    suggestionText,
                    trailing,
                    insertionStrategy,
                    insertionRange,
                    cursorBeforeTrailing,
                    escapeSuggestion
                  );
                },
                "data-dt-name": "Suggestion",
                children: [
                  /* @__PURE__ */ jsxs(Flex, { gap: 8, minWidth: 0, children: [
                    icon && /* @__PURE__ */ jsx(Chip, { size: "condensed", className: styles.icon, children: /* @__PURE__ */ jsx(Chip.Prefix, { children: icon }) }),
                    /* @__PURE__ */ jsx(
                      Flex,
                      {
                        className: clsx(styles.value, {
                          [`${styles.truncateWithinMark}`]: id === "strato-filterfield-suggestion-use-as-value" || id === "strato-filterfield-suggestion-use-as-key" || id === "strato-filterfield-suggestion-pasted-content" || id === "strato-filterfield-suggestion-search"
                        }),
                        ref: valueRef,
                        children: /* @__PURE__ */ jsx(WrappedSuggestionLabel, { displayLabel })
                      }
                    )
                  ] }),
                  finalDetails
                ]
              }
            )
          }
        ),
        slots.pin
      ]
    }
  );
});
Object.assign(Suggestion, { displayValue: "FilterField.Suggestion" });
export {
  Suggestion
};
