// @dynatrace/strato-components/esm/filters/filter-field/suggestions/SuggestionPin.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import {
  useContext,
  useEffect,
  useRef,
  useState
} from "react";
import { useIntl } from "react-intl";
import { PinIcon, PinOffIcon } from "@dynatrace/strato-icons";
import { Button } from "../../../buttons/button/Button.js";
import { getVirtualFocusStyles } from "../../../core/styles/getVirtualFocusStyles.js";
import { virtualFocus } from "../../../core/styles/virtualFocus.sty.js";
import { Tooltip } from "../../../overlays/tooltip/Tooltip.js";
import { FilterFieldContext } from "../contexts/FilterFieldContext.js";
import {
  usePinnedFilterMutation,
  usePinnedFilters
} from "../hooks/recent-and-pinned/usePinnedFilters.js";
import { useRecentAndPinnedFiltersStore } from "../hooks/recent-and-pinned/useRecentFiltersStore.js";
import { useFilterFieldContext } from "../hooks/useFilterFieldContext.js";
import { filterFieldMessages } from "../messages.js";
function useSuggestionPin(filterString) {
  const { filterNamespace } = useContext(FilterFieldContext);
  const pinnedFilters = usePinnedFilters(filterNamespace);
  const { pinFilter, unpinFilter } = usePinnedFilterMutation(filterNamespace);
  const isPinned = (pinnedFilters ?? []).find(
    (storedFilter) => storedFilter.filter === filterString.trim()
  ) !== void 0;
  return {
    isPinned,
    pinFilter,
    unpinFilter,
    limitReached: pinnedFilters.length >= 10
  };
}
function SuggestionPin(props) {
  const {
    id,
    filterString,
    className,
    size,
    placement,
    fallbackPlacements,
    role,
    disabled,
    "aria-disabled": ariaDisabledProps,
    showError,
    ...remainingProps
  } = props;
  const intl = useIntl();
  const { focusedListItem, filterNamespace } = useFilterFieldContext();
  const canPin = useRecentAndPinnedFiltersStore(
    (state) => state.canPin,
    filterNamespace
  );
  const { isPinned, pinFilter, unpinFilter, limitReached } = useSuggestionPin(filterString);
  const pinUnpinTooltip = intl.formatMessage(
    filterFieldMessages.suggestions.pinUnpinTooltip,
    { isPinned }
  );
  const pinUnpinLabel = intl.formatMessage(
    filterFieldMessages.suggestions.pinUnpinLabel,
    { isPinned, filterString }
  );
  const disabledTooltip = intl.formatMessage(
    filterFieldMessages.suggestions.pinUnpinLimitTooltip
  );
  const ariaDisabled = limitReached && !isPinned;
  const buttonRef = useRef(null);
  const [tooltipOpen, setTooltipOpen] = useState(false);
  useEffect(() => {
    const classListObserver = new MutationObserver((entries) => {
      for (const entry of entries) {
        if (entry.target === buttonRef.current && entry.target instanceof HTMLElement) {
          if (entry.target.classList.contains(virtualFocus.neutral)) {
            setTooltipOpen(true);
          } else {
            setTooltipOpen(false);
          }
        }
      }
    });
    if (buttonRef.current) {
      classListObserver.observe(buttonRef.current, {
        attributes: true,
        attributeFilter: ["class"]
      });
    }
    return () => {
      classListObserver.disconnect();
    };
  }, []);
  if (!canPin) {
    return null;
  }
  return /* @__PURE__ */ jsx(
    Tooltip,
    {
      open: !disabled && tooltipOpen,
      onOpenChange: setTooltipOpen,
      placement,
      fallbackPlacements,
      text: ariaDisabled ? disabledTooltip : pinUnpinTooltip,
      children: /* @__PURE__ */ jsx(
        Button,
        {
          size,
          role,
          ref: buttonRef,
          id,
          disabled,
          ...remainingProps,
          "data-suggestion-pin": true,
          ...role === "option" ? {
            "aria-selected": focusedListItem?.id === id,
            "aria-checked": isPinned
          } : { "aria-pressed": isPinned },
          "aria-label": ariaDisabled ? disabledTooltip : pinUnpinLabel,
          "aria-disabled": ariaDisabledProps || ariaDisabled,
          color: showError ? "critical" : "neutral",
          className: clsx(
            className,
            getVirtualFocusStyles(focusedListItem?.id === id).className
          ),
          onPointerDown: (event) => {
            event.preventDefault();
            event.stopPropagation();
            event.nativeEvent.stopImmediatePropagation();
          },
          onClick: () => {
            if (ariaDisabled) {
              return;
            }
            if (isPinned) {
              unpinFilter(filterString);
            } else {
              pinFilter(filterString);
            }
          },
          children: isPinned ? /* @__PURE__ */ jsx(PinOffIcon, {}) : /* @__PURE__ */ jsx(PinIcon, {})
        }
      )
    }
  );
}
export {
  SuggestionPin,
  useSuggestionPin
};
