// @dynatrace/strato-components/esm/filters/filter-field/suggestions/Suggestion.sty.js
import "./SuggestionsOverlay.css";
import "../../../core/components/app-root/AppRoot.css";
import "../../../core/styles/virtualFocus.css";
import "./Suggestion.css";
var defaultSuggestion = "_1smjoa1c-3-6-0";
var icon = "_1smjoa10-3-6-0";
var matchableSuggestionDetails = "_1smjoa12-3-6-0";
var pastingSuggestion = "_1smjoa1f-3-6-0";
var searchSuggestion = "_1smjoa1d-3-6-0";
var statementSuggestionValue = "_1smjoa1e-3-6-0";
var suggestion = "_1smjoa14-3-6-0";
var suggestionContent = "_1smjoa15-3-6-0";
var suggestionText = "_1smjoa16-3-6-0";
var suggestionsEmptyState = "_1smjoa17-3-6-0";
var suggestionsJSONPathHint = "_1smjoa18-3-6-0";
var suggestionsLoading = "_1smjoa19-3-6-0";
var tooltipWrapper = "_1smjoa1b-3-6-0";
var truncateWithinMark = "_1smjoa1a-3-6-0";
var value = "_1smjoa11-3-6-0";
var virtualizedSuggestion = "_1smjoa13-3-6-0";
export {
  defaultSuggestion,
  icon,
  matchableSuggestionDetails,
  pastingSuggestion,
  searchSuggestion,
  statementSuggestionValue,
  suggestion,
  suggestionContent,
  suggestionText,
  suggestionsEmptyState,
  suggestionsJSONPathHint,
  suggestionsLoading,
  tooltipWrapper,
  truncateWithinMark,
  value,
  virtualizedSuggestion
};
