// @dynatrace/strato-components/esm/filters/filter-field/suggestions/Suggestions.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { useVirtualizer } from "@tanstack/react-virtual";
import clsx from "clsx";
import { motion } from "framer-motion";
import {
  Children,
  cloneElement,
  isValidElement,
  useMemo,
  useRef
} from "react";
import { FormattedMessage, useIntl } from "react-intl";
import { usePopper } from "react-popper";
import useOverlayTransition from "../hooks/useOverlayTransition.js";
import {
  sliceFlattenedSuggestions,
  useShowMore
} from "../hooks/useShowMore.js";
import { filterFieldMessages } from "../messages.js";
import { AriaLiveInfo } from "./AriaLiveInfo.js";
import { FallbackKeySuggestionGroup } from "./FallbackKeySuggestions.js";
import { LogicalOperatorSuggestionsGroup } from "./LogicalOperatorSuggestions.js";
import { PastedContentSuggestionGroup } from "./PastedContentSuggestions.js";
import {
  PinnedSuggestionGroup,
  RecentSuggestionGroup
} from "./RecentPinnedSuggestionItem.js";
import { SearchSuggestionGroup } from "./SearchSuggestions.js";
import * as suggestionStyles from "./Suggestion.sty.js";
import { Suggestion } from "./Suggestion.js";
import * as suggestionGroupStyles from "./SuggestionGroup.sty.js";
import { SuggestionGroupLabel } from "./SuggestionGroupLabel.js";
import * as styles from "./SuggestionsOverlay.sty.js";
import { Button } from "../../../buttons/button/Button.js";
import { ProgressCircle } from "../../../content/progress/ProgressCircle.js";
import { VisuallyHidden } from "../../../core/components/visually-hidden/VisuallyHidden.js";
import { useBehavioralTrackingProps } from "../../../core/hooks/useBehavioralTrackingProps.js";
import { Flex } from "../../../layouts/flex/Flex.js";
import { Surface } from "../../../layouts/surface/Surface.js";
import { CSSHighlight } from "../components/CSSHighlight.js";
import { useFilterFieldContext } from "../hooks/useFilterFieldContext.js";
import { flattenSuggestions } from "../utils/suggestions-utils/flatten-suggestions.js";
function toGroupLabelId(groupId) {
  return `group-label-${groupId.replace(/\s+/g, "-")}`;
}
function renderFlattenedSuggestionItem(flatItem, trailing) {
  switch (flatItem.kind) {
    case "group-header": {
      let child;
      if (isValidElement(flatItem.label)) {
        child = flatItem.label;
      } else if (flatItem.label) {
        child = /* @__PURE__ */ jsx(SuggestionGroupLabel, { children: flatItem.label });
      } else {
        child = null;
      }
      return {
        child: /* @__PURE__ */ jsx(
          "div",
          {
            className: suggestionGroupStyles.groupHeaderSeparator,
            role: "presentation",
            children: child
          }
        ),
        key: `group-header-${flatItem.groupId}`
      };
    }
    case "suggestion": {
      const ariaDescribedBy = flatItem.groupId ? toGroupLabelId(flatItem.groupId) : void 0;
      return {
        child: /* @__PURE__ */ jsx(
          Suggestion,
          {
            ...flatItem.data,
            trailing: flatItem.data.trailing ?? trailing,
            "aria-describedby": ariaDescribedBy
          }
        ),
        key: `${flatItem.data.value ?? flatItem.data.displayValue ?? ""}`
      };
    }
    case "react-node": {
      const groupLabelId = flatItem.groupId ? toGroupLabelId(flatItem.groupId) : void 0;
      let child = flatItem.node;
      let key = "";
      if (isValidElement(flatItem.node) && groupLabelId) {
        const existing = flatItem.node.props["aria-describedby"] ?? "";
        const merged = existing ? `${existing} ${groupLabelId}` : groupLabelId;
        child = cloneElement(flatItem.node, {
          "aria-describedby": merged
        });
        key = `react-node-${flatItem.node.key}`;
      }
      return { child, key };
    }
  }
}
function SuggestionContent(props) {
  const { optionVirtualizer, flattenedSuggestions, isSliced, controlProps } = props;
  const intl = useIntl();
  const {
    term,
    editorView,
    autoSuggestions,
    recentSuggestionValues,
    pinnedSuggestionValues,
    logicalOperatorSuggestionValues,
    pastedContentSuggestionValues,
    searchSuggestionValues,
    fallbackKeySuggestionValues,
    statementSuggestionValues,
    filterFieldId,
    parserConfig,
    isValidSearch
  } = useFilterFieldContext();
  const { trailing } = autoSuggestions;
  const showLogicalOperators = logicalOperatorSuggestionValues.length > 0;
  const showRecents = recentSuggestionValues.length > 0;
  const showPinned = pinnedSuggestionValues.length > 0;
  const showFallbackKeySuggestions = fallbackKeySuggestionValues.length > 0;
  const showStatementSuggestions = statementSuggestionValues.length > 0;
  const showSearchSuggestions = isValidSearch?.current && searchSuggestionValues.length > 0 && // eslint-disable-next-line @typescript-eslint/no-deprecated
  (parserConfig?.searchConversion ?? false);
  const showPastedContentSuggestions = pastedContentSuggestionValues.length > 0;
  const showBuiltInSuggestions = showRecents || showPinned || showLogicalOperators || showPastedContentSuggestions || showSearchSuggestions || showFallbackKeySuggestions || showStatementSuggestions;
  const isEmpty = editorView?.state.doc.toString() === "";
  const childrenCount = flattenedSuggestions.length;
  if (childrenCount === 0 && !showBuiltInSuggestions) {
    return /* @__PURE__ */ jsx(
      Suggestion,
      {
        trailing: "",
        displayValue: intl.formatMessage(
          filterFieldMessages.suggestions.noSuggestions
        )
      }
    );
  }
  return /* @__PURE__ */ jsxs("div", { id: `suggestions-${filterFieldId}`, children: [
    childrenCount > 0 ? /* @__PURE__ */ jsx(
      CSSHighlight,
      {
        term: term ?? "",
        selector: `.${suggestionStyles.value}:not(:has(mark)), .${suggestionStyles.matchableSuggestionDetails}:not(:has(mark))`,
        highlightName: "filterfield-suggestions-highlight",
        children: /* @__PURE__ */ jsx(
          "div",
          {
            style: {
              width: "100%",
              position: "relative",
              height: optionVirtualizer.getTotalSize(),
              display: "grid",
              gridTemplateRows: optionVirtualizer.measurementsCache.map((el) => `${el.size}px`).join(" ")
            },
            children: optionVirtualizer.getVirtualItems().map((virtualRow) => {
              const flatItem = flattenedSuggestions[virtualRow.index];
              const { child, key: suggestionKey } = renderFlattenedSuggestionItem(flatItem, trailing);
              return /* @__PURE__ */ jsx(
                "div",
                {
                  className: clsx(
                    "strato-filter-field-suggestion",
                    suggestionStyles.virtualizedSuggestion
                  ),
                  "data-index": virtualRow.index,
                  style: {
                    gridRow: virtualRow.index + 1,
                    minWidth: 0,
                    height: "fit-content"
                  },
                  ref: (el) => {
                    optionVirtualizer.measureElement(el);
                  },
                  children: child
                },
                `${virtualRow.key}-${suggestionKey}`
              );
            })
          },
          `wrapper-${childrenCount}`
        )
      }
    ) : null,
    isSliced && /* @__PURE__ */ jsx("div", { role: "presentation", children: /* @__PURE__ */ jsx(Button, { ...controlProps }) }),
    showBuiltInSuggestions && /* @__PURE__ */ jsxs(Fragment, { children: [
      /* @__PURE__ */ jsx(PastedContentSuggestionGroup, {}),
      /* @__PURE__ */ jsx(FallbackKeySuggestionGroup, {}),
      /* @__PURE__ */ jsx(LogicalOperatorSuggestionsGroup, {}),
      showSearchSuggestions && !isEmpty && /* @__PURE__ */ jsx(SearchSuggestionGroup, {}),
      /* @__PURE__ */ jsxs(
        CSSHighlight,
        {
          className: suggestionGroupStyles.builtInSuggestionsContainer,
          term: editorView?.state.doc.toString() ?? "",
          selector: `.${suggestionStyles.value}`,
          highlightName: "filterfield-suggestions-highlight",
          children: [
            /* @__PURE__ */ jsx(RecentSuggestionGroup, {}),
            /* @__PURE__ */ jsx(PinnedSuggestionGroup, {})
          ]
        }
      ),
      showSearchSuggestions && isEmpty && /* @__PURE__ */ jsx(SearchSuggestionGroup, {})
    ] })
  ] });
}
const heightModifier = {
  name: "filterfield-height-handler",
  enabled: true,
  fn: ({ state }) => {
    const viewportHeight = Math.max(
      document.documentElement.clientHeight || 0,
      window.innerHeight || 0
    );
    const maxHeightBase = Math.max(viewportHeight / 2, 320);
    state.styles.popper.maxHeight = `${maxHeightBase - state.rects.reference.height - 20}px`;
  },
  phase: "beforeWrite",
  requires: ["computeStyles"]
};
const suggestionsAnimationVariants = {
  inital: {
    translateY: "-10px",
    // Compensate the parent scaling the content to 0.8. Due to round errors when the height gets taller
    // the content sometimes got larger than the intended which lead to a "vertigo" effect. By adding
    // 0.05 we can counteract that effect somewhat.
    scaleY: 1 / 0.85,
    opacity: 0
  },
  default: {
    translateY: 0,
    scaleY: 1,
    opacity: 1
  },
  exit: {
    translateY: "-5px",
    opacity: 0
  }
};
function getSuggestionsCount(defaultSuggestionsCount, isEmpty) {
  if (typeof defaultSuggestionsCount === "number") {
    return defaultSuggestionsCount;
  }
  return defaultSuggestionsCount[isEmpty ? "empty" : "filled"];
}
const Suggestions = (props) => {
  const {
    loading,
    children,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps] = useBehavioralTrackingProps(remainingProps);
  const { autoSuggestions, defaultSuggestionsCount, triggerRef, editorView } = useFilterFieldContext();
  const intl = useIntl();
  const transition = useOverlayTransition(true);
  const parentRef = useRef(null);
  const childrenArray = useMemo(() => Children.toArray(children), [children]);
  const { flattenedSuggestionItems, groupMetadata } = useMemo(
    () => flattenSuggestions([
      ...Object.values(autoSuggestions.groupedSuggestions ?? {}).filter(
        (group) => group && (group.suggestions?.length ?? 0) > 0
      ),
      ...childrenArray
    ]),
    [autoSuggestions.groupedSuggestions, childrenArray]
  );
  const isEmpty = editorView?.state.doc.toString() === "";
  const suggestionsCount = getSuggestionsCount(
    defaultSuggestionsCount,
    isEmpty
  );
  const {
    items: slicedSuggestions,
    isSliced,
    controlProps
  } = useShowMore({
    items: flattenedSuggestionItems,
    primaryCount: suggestionsCount,
    sliceFn: sliceFlattenedSuggestions
  });
  const optionVirtualizer = useVirtualizer({
    count: slicedSuggestions.length,
    getScrollElement: () => parentRef.current,
    overscan: 5,
    estimateSize: () => 32
  });
  const suggestionsRef = useRef(null);
  const { styles: positioningStyles } = usePopper(
    triggerRef,
    suggestionsRef.current,
    {
      placement: "bottom-start",
      modifiers: [heightModifier]
    }
  );
  return /* @__PURE__ */ jsxs(
    Surface,
    {
      className: clsx(consumerClassName, styles.suggestionsOverlay),
      style: consumerStyle,
      elevation: "floating",
      ref: suggestionsRef,
      "data-testid": dataTestId,
      ...behavioralTrackingProps,
      children: [
        /* @__PURE__ */ jsx(AriaLiveInfo, { flattenedSuggestionItems }),
        groupMetadata.length > 0 && /* @__PURE__ */ jsx(VisuallyHidden, { children: groupMetadata.map(({ groupId, label }) => /* @__PURE__ */ jsx("span", { id: toGroupLabelId(groupId), children: label }, groupId)) }),
        /* @__PURE__ */ jsx(
          "div",
          {
            className: styles.suggestionsScrollContainerWrapper,
            role: "listbox",
            "aria-expanded": "true",
            "aria-label": intl.formatMessage(
              filterFieldMessages.suggestions.a11y.ariaLabel
            ),
            children: /* @__PURE__ */ jsx(
              motion.div,
              {
                className: clsx(
                  styles.suggestionsScrollContainer,
                  "strato-filter-field-suggestions"
                ),
                ref: parentRef,
                style: { maxHeight: positioningStyles.popper.maxHeight },
                variants: suggestionsAnimationVariants,
                initial: "inital",
                animate: "default",
                exit: "exit",
                transition,
                children: loading ? /* @__PURE__ */ jsxs(
                  Flex,
                  {
                    role: "presentation",
                    justifyContent: "start",
                    px: 8,
                    py: 6,
                    className: suggestionStyles.suggestionsLoading,
                    children: [
                      /* @__PURE__ */ jsx(ProgressCircle, { size: "small", "aria-label": "Loading suggestions" }),
                      /* @__PURE__ */ jsx(FormattedMessage, { ...filterFieldMessages.suggestions.loading })
                    ]
                  }
                ) : /* @__PURE__ */ jsx(
                  SuggestionContent,
                  {
                    optionVirtualizer,
                    flattenedSuggestions: slicedSuggestions,
                    isSliced,
                    controlProps
                  }
                )
              }
            )
          }
        )
      ]
    }
  );
};
Suggestions.displayValue = "FilterField.Suggestions";
export {
  Suggestions
};
