// @dynatrace/strato-components/esm/filters/filter-field/suggestions/SuggestionTips.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import { motion } from "framer-motion";
import { useIntl } from "react-intl";
import * as styles from "./SuggestionTips.sty.js";
import { Flex } from "../../../layouts/flex/Flex.js";
import { Surface } from "../../../layouts/surface/Surface.js";
import { textStyle as textStyleCSS } from "../../../styles/textStyle.sty.js";
import { ExternalLink } from "../../../typography/external-link/ExternalLink.js";
import { Text } from "../../../typography/text/Text.js";
import { TextEllipsis } from "../../../typography/text-ellipsis/TextEllipsis.js";
import useOverlayTransition from "../hooks/useOverlayTransition.js";
import { filterFieldMessages } from "../messages.js";
const suggestionTipsVariants = {
  enter: {
    // Compensate the parent scaling the content to 0.8
    scaleY: 1 / 0.85,
    opacity: 0
  },
  default: {
    scaleY: 1,
    opacity: 1
  },
  exit: {
    opacity: 0
  }
};
const SuggestionTips = () => {
  const intl = useIntl();
  const transition = useOverlayTransition(true);
  const tips = [
    {
      shortcut: "\u23CE",
      text: intl.formatMessage(filterFieldMessages.suggestions.tips.apply)
    },
    {
      shortcut: "\u2193\u2191",
      text: intl.formatMessage(filterFieldMessages.suggestions.tips.selection)
    },
    {
      shortcut: "esc",
      text: intl.formatMessage(filterFieldMessages.suggestions.tips.close)
    }
  ];
  return /* @__PURE__ */ jsx(
    Surface,
    {
      className: styles.suggestionTips,
      paddingX: 16,
      paddingY: 8,
      elevation: "floating",
      children: /* @__PURE__ */ jsxs(
        Flex,
        {
          as: motion.div,
          className: styles.suggestionTipsWrapper,
          columnGap: 16,
          rowGap: 8,
          justifyContent: "space-between",
          flexWrap: "wrap",
          variants: suggestionTipsVariants,
          initial: "enter",
          animate: "default",
          exit: "exit",
          transition,
          children: [
            /* @__PURE__ */ jsx(
              Flex,
              {
                className: styles.suggestionTipsContent,
                style: { overflow: "hidden" },
                columnGap: 16,
                rowGap: 8,
                children: tips.map((tip) => {
                  return /* @__PURE__ */ jsxs(Flex, { maxWidth: "100%", gap: 4, children: [
                    /* @__PURE__ */ jsxs(Text, { textStyle: "small-emphasized", children: [
                      tip.shortcut,
                      " "
                    ] }),
                    /* @__PURE__ */ jsx(
                      TextEllipsis,
                      {
                        className: clsx(
                          textStyleCSS({ textStyle: "small" }),
                          styles.suggstionTipsEllipsis
                        ),
                        children: tip.text
                      }
                    )
                  ] }, tip.shortcut + tip.text);
                })
              }
            ),
            /* @__PURE__ */ jsx(
              ExternalLink,
              {
                href: "https://docs.dynatrace.com/docs/get-started/dynatrace-ui/ui-filter-field",
                className: clsx(
                  textStyleCSS({ textStyle: "small" }),
                  styles.docsLink
                ),
                "data-dt-name": "FilterSyntax",
                children: intl.formatMessage(filterFieldMessages.docsLink)
              }
            )
          ]
        }
      )
    }
  );
};
export {
  SuggestionTips
};
