// @dynatrace/strato-components/filters/segment-selector/parts/ShowMorePanel/ShowMorePanel.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ShowMorePanel_exports = {};
__export(ShowMorePanel_exports, {
  ShowMorePanel: () => ShowMorePanel
});
module.exports = __toCommonJS(ShowMorePanel_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_react_is = require("react-is");
var import_strato_icons = require("@dynatrace/strato-icons");
var styles = __toESM(require("./ShowMorePanel.sty.js"));
var import_Button = require("../../../../buttons/button/Button.js");
var import_is_node_environment = require("../../../../core/utils/is-node-environment.js");
var import_Flex = require("../../../../layouts/flex/Flex.js");
const ExtendedContent = ({ children }) => children;
const PrimaryContent = ({ children }) => children;
function useChildrenSlots(children) {
  const memoizedSlots = (0, import_react.useMemo)(() => {
    const slots = {};
    import_react.Children.forEach(children, (child) => {
      if ((0, import_react_is.isElement)(child)) {
        switch (child.type) {
          case ExtendedContent:
            slots.extendedContent = child;
            break;
          case PrimaryContent:
          default:
            slots.primaryContent = child;
            break;
        }
      }
    });
    return slots;
  }, [children]);
  return memoizedSlots;
}
const ShowMorePanelRoot = ({ children }) => {
  const { primaryContent, extendedContent } = useChildrenSlots(children);
  const [isExpanded, setIsExpanded] = (0, import_react.useState)(false);
  if (!children) {
    if ((0, import_is_node_environment.isDev)()) {
      console.warn(
        "SegmentSelector: The used child is not supported by the ExpandableSection."
      );
    }
    return null;
  }
  if (!extendedContent) {
    return primaryContent;
  }
  const icon = isExpanded ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ChevronUpSmallIcon, {}) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ChevronDownSmallIcon, {});
  const dataDtProperties = `state:${isExpanded ? "less" : "more"}`;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)("ul", { className: styles.list, children: primaryContent }),
    isExpanded && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "ul",
      {
        className: (0, import_clsx.default)(styles.expandableWrapper, {
          [styles.expandableWrapperExpanded]: isExpanded,
          [styles.expandableWrapperCollapsed]: !isExpanded
        }),
        children: extendedContent
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Button.Button,
      {
        onClick: () => setIsExpanded((previousState) => !previousState),
        width: "full",
        textAlign: "start",
        style: {
          transitionTimingFunction: "none",
          transitionDuration: "0s",
          transition: "none"
        },
        "data-dt-name": "ShowMoreLess",
        "data-dt-properties": dataDtProperties,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { gap: 2, as: "span", paddingLeft: 4, children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_react_intl.FormattedMessage,
            {
              defaultMessage: "Show {isExpanded, select, true {less} false {more} other{}}",
              id: "mYmd/xMZulbADHsY",
              description: "Text shown on the expanding button within the SegmentSelector preset list, when currently expanded.",
              values: { isExpanded }
            }
          ),
          icon
        ] })
      }
    )
  ] });
};
const ShowMorePanel = Object.assign(ShowMorePanelRoot, {
  PrimaryContent,
  ExtendedContent
});
