// @dynatrace/strato-components/esm/overlays/modal/BaseModal.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import {
  AnimatePresence,
  motion,
  useReducedMotion
} from "framer-motion";
import {
  forwardRef,
  useMemo,
  useRef
} from "react";
import AnimationsJS from "@dynatrace/strato-design-tokens/animations-js";
import * as styles from "./Modal.sty.js";
import { FocusScope } from "../../core/components/focus-scope/FocusScope.js";
import { OverlayContainer } from "../../core/components/OverlayContainer.js";
import { useOverlayStacking } from "../../core/hooks/useOverlayStacking.js";
import { OverlayStackingProvider } from "../../core/providers/OverlayStackingProvider.js";
import {
  fadeInEasing,
  fadeOutEasing
} from "../../core/utils/framer-motion-easings.js";
import { useDialog } from "../hooks/useDialog.js";
import { useOverlay } from "../hooks/useOverlay.js";
const BaseModal = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const {
    show = false,
    size = "medium",
    children,
    onDismiss,
    dismissible = true,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    className: consumerClassName,
    style: consumerStyle,
    "data-dt-component": dtComponentName,
    ...remainingProps
  } = props;
  const ref = useRef(null);
  const { overlayProps } = useOverlay(
    {
      ...props,
      onClose: () => {
        onDismiss?.();
      },
      isOpen: show,
      isKeyboardDismissDisabled: !dismissible
    },
    ref
  );
  const { dialogProps, titleProps } = useDialog(remainingProps);
  const overlayContainerProps = useOverlayStacking(void 0, "overlay");
  const reducedMotion = useReducedMotion();
  const animationVariants = useMemo(() => {
    const duration = reducedMotion ? 0 : AnimationsJS.FadeIn.Duration.Fast;
    return {
      // When animating AnimatePresence where child components
      // will also need to transform, they can share the same enter/exit variant states.
      // The keys will need to be exactly the same, and the child will not need to
      // define `initial`, `animate` or `exit` variants on the component.
      backdrop: {
        enter: {
          // For the opacity we need to use the default value here. Framer motion would complain
          // about too complex values, when transitioning from two custom property values within
          // an rgba value apparently.
          opacity: 1,
          transition: {
            duration,
            ease: fadeInEasing
          }
        },
        exit: {
          opacity: 0,
          transition: {
            duration,
            ease: fadeOutEasing
          }
        }
      },
      modal: {
        enter: {
          y: 0,
          scale: 1,
          transition: {
            duration,
            ease: fadeInEasing
          }
        },
        exit: {
          y: "5px",
          scale: 0.98,
          transition: {
            duration,
            ease: fadeInEasing
          }
        }
      }
    };
  }, [reducedMotion]);
  return /* @__PURE__ */ jsx(
    OverlayContainer,
    {
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...overlayContainerProps,
      "data-dt-component": dtComponentName,
      children: /* @__PURE__ */ jsx(AnimatePresence, { children: show && /* @__PURE__ */ jsx(
        motion.div,
        {
          className: styles.backdrop,
          variants: animationVariants.backdrop,
          initial: "exit",
          animate: "enter",
          exit: "exit",
          ref: forwardedRef,
          onPointerDown: (event) => {
            if (event.target === event.currentTarget && dismissible) {
              onDismiss?.();
            }
          },
          children: /* @__PURE__ */ jsx(FocusScope, { contain: true, autoFocus: true, children: /* @__PURE__ */ jsx(
            motion.div,
            {
              ...overlayProps,
              ...dialogProps,
              ...titleProps,
              className: clsx(
                styles.base,
                styles.sizes[size],
                consumerClassName
              ),
              style: consumerStyle,
              ref,
              variants: animationVariants.modal,
              ...remainingProps,
              children: /* @__PURE__ */ jsx(OverlayStackingProvider, { children })
            }
          ) })
        }
      ) })
    }
  );
});
BaseModal.displayName = "BaseModal";
export {
  BaseModal
};
