// @dynatrace/strato-components/esm/overlays/modal/Modal.js
import { jsx, jsxs } from "react/jsx-runtime";
import { forwardRef } from "react";
import { useIntl } from "react-intl";
import { XmarkIcon } from "@dynatrace/strato-icons";
import { BaseModal } from "./BaseModal.js";
import * as styles from "./Modal.sty.js";
import { Button } from "../../buttons/button/Button.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useId } from "../../core/hooks/useId.js";
import { canReceiveDefaultTextFormat as _canReceiveDefaultTextFormat } from "../../core/utils/can-receive-default-text-format.js";
import { Divider } from "../../layouts/divider/Divider.js";
import { globalSprinkles } from "../../styles/safe-sprinkles.js";
import { Heading } from "../../typography/heading/Heading.js";
const COMPONENT_NAME = "Modal";
function getVisualHeadingLevelForSize(size) {
  switch (size) {
    case "small":
      return 5;
    case "medium":
    case "large":
      return 4;
    default:
      return 4;
  }
}
const Modal = /* @__PURE__ */ forwardRef(
  (props, forwardedRef) => {
    const {
      show = false,
      title,
      footer,
      size = "medium",
      children,
      onDismiss,
      dismissible = true,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      className: consumerClassName,
      style: consumerStyle,
      ...remainingProps
    } = props;
    const visualHeadingSize = getVisualHeadingLevelForSize(size);
    const intl = useIntl();
    const headerId = useId();
    const closeButtonLabel = intl.formatMessage({
      defaultMessage: "Close modal",
      id: "aB3uCeB71q6UURBy",
      description: "Aria-label for close modal button."
    });
    const hasCustomTitle = !_canReceiveDefaultTextFormat(title);
    const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
    return /* @__PURE__ */ jsxs(
      BaseModal,
      {
        "aria-labelledby": headerId,
        show,
        size,
        onDismiss,
        dismissible,
        ref: forwardedRef,
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        className: consumerClassName,
        style: consumerStyle,
        ...restProps,
        ...behavioralTrackingProps,
        "data-dt-component": COMPONENT_NAME,
        children: [
          /* @__PURE__ */ jsxs("header", { className: styles.header, children: [
            hasCustomTitle ? /* @__PURE__ */ jsx("div", { style: { display: "contents" }, id: headerId, children: title }) : /* @__PURE__ */ jsx(
              Heading,
              {
                id: headerId,
                level: visualHeadingSize,
                as: "h2",
                style: {
                  wordBreak: "normal",
                  overflowWrap: "anywhere"
                },
                children: title
              }
            ),
            dismissible && /* @__PURE__ */ jsx(
              Button,
              {
                style: { alignSelf: "start" },
                onClick: onDismiss,
                "aria-label": closeButtonLabel,
                variant: "default",
                "data-dt-name": "CloseX",
                children: /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(XmarkIcon, {}) })
              }
            )
          ] }),
          (children || footer) && /* @__PURE__ */ jsx(
            "div",
            {
              className: globalSprinkles({
                paddingBottom: 0,
                paddingTop: 0,
                paddingLeft: 24,
                paddingRight: 24
              }),
              children: /* @__PURE__ */ jsx(Divider, {})
            }
          ),
          children && /* @__PURE__ */ jsx(
            "div",
            {
              className: globalSprinkles({
                paddingTop: 12,
                paddingRight: 24,
                paddingLeft: 24,
                paddingBottom: footer !== void 0 ? 12 : 24
              }),
              style: { overflowY: "auto" },
              children
            }
          ),
          footer && /* @__PURE__ */ jsx(
            "footer",
            {
              className: globalSprinkles({
                display: "flex",
                alignItems: "flex-end",
                justifyContent: "flex-end",
                paddingBottom: 24,
                paddingTop: 12,
                paddingLeft: 24,
                paddingRight: 24
              }),
              children: footer
            }
          )
        ]
      }
    );
  }
);
Modal.displayName = COMPONENT_NAME;
export {
  Modal
};
