// @dynatrace/strato-components/charts/core/components/series-action/ChartSeriesAction.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ChartSeriesAction_exports = {};
__export(ChartSeriesAction_exports, {
  ChartSeriesAction: () => ChartSeriesAction
});
module.exports = __toCommonJS(ChartSeriesAction_exports);
var import_jsx_runtime = (
  // eslint-disable-next-line react/jsx-no-useless-fragment
  require("react/jsx-runtime")
);
var import_Group = require("../../../../navigation/menu/Group.js");
var import_Intent = require("../../../../navigation/menu/Intent.js");
var import_Item = require("../../../../navigation/menu/Item.js");
var import_ItemTooltip = require("../../../../navigation/menu/ItemTooltip.js");
var import_Label = require("../../../../navigation/menu/Label.js");
var import_Link = require("../../../../navigation/menu/Link.js");
var import_Prefix = require("../../../../navigation/menu/Prefix.js");
var import_Sub = require("../../../../navigation/menu/Sub.js");
var import_SubContent = require("../../../../navigation/menu/SubContent.js");
var import_SubTrigger = require("../../../../navigation/menu/SubTrigger.js");
const ChartSeriesActionRoot = ({
  children
}) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, { children });
ChartSeriesActionRoot["displayName"] = "ChartSeriesAction";
const ChartSeriesAction = Object.assign(ChartSeriesActionRoot, {
  Intent: ({ icon, ...props }) => /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Intent.Intent, { ...props, children: [
    icon && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Prefix.Prefix, { children: icon }),
    props.children
  ] }),
  Item: import_Item.Item,
  ItemTooltip: import_ItemTooltip.ItemTooltip,
  Label: import_Label.Label,
  ItemIcon: import_Prefix.Prefix,
  Group: import_Group.Group,
  Sub: import_Sub.Sub,
  SubTrigger: import_SubTrigger.SubTrigger,
  SubContent: import_SubContent.SubContent,
  Link: (props) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Link.Link, { "data-link": true, ...props })
});
