// @dynatrace/strato-components/esm/charts/core/components/overlay-tooltip/contexts/TooltipContext.js
import { jsx } from "react/jsx-runtime";
import {
  createContext,
  useContext,
  useRef
} from "react";
import { createStore } from "../../../../../core/store/store.js";
import { emptyData } from "../utils.js";
const initialState = () => ({
  pinned: false,
  inBounds: false,
  activeSeries: null,
  absoluteX: null,
  absoluteY: null,
  relativeX: null,
  relativeY: null,
  percentX: null,
  percentY: null,
  data: emptyData()
});
const TooltipContext = createContext({});
const useTooltip = () => {
  const tooltip = useContext(TooltipContext);
  if (!tooltip) {
    throw new Error("useTooltip not used within TooltipContext provider.");
  }
  return tooltip;
};
function TooltipProvider(props) {
  const store = useRef();
  if (props.persistedStore) {
    store.current ||= createStore(initialState());
  } else {
    store.current = createStore(initialState());
  }
  return /* @__PURE__ */ jsx(TooltipContext.Provider, { value: store.current, children: props.children });
}
export {
  TooltipContext,
  TooltipProvider,
  useTooltip
};
