// @dynatrace/strato-components/charts/core/components/canvas/Canvas.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Canvas_exports = {};
__export(Canvas_exports, {
  Canvas: () => Canvas
});
module.exports = __toCommonJS(Canvas_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_CanvasContext = require("./CanvasContext.js");
const setupCanvas = (canvasElement, width, height) => {
  const dpr = Math.max(1, window.devicePixelRatio) || 1;
  canvasElement.width = width * dpr;
  canvasElement.height = height * dpr;
  const context = canvasElement.getContext("2d");
  context?.scale(dpr, dpr);
  canvasElement.style.width = "100%";
  canvasElement.style.height = "100%";
  return context;
};
const draw = (canvasElement, context, shapeRefs) => {
  if (!canvasElement) {
    return;
  }
  if (context) {
    context.clearRect(0, 0, canvasElement.width, canvasElement.height);
    shapeRefs.forEach((ref) => {
      ref.current.draw(context, ref.current.id);
    });
  }
};
function ensureCanvasContext(currentCtx, canvasRef, width, height) {
  if (!currentCtx.ctx || currentCtx.prevHeight !== height || currentCtx.prevWidth !== width) {
    if (canvasRef.current) {
      currentCtx.ctx = setupCanvas(canvasRef.current, width, height);
    } else {
      currentCtx.ctx = null;
    }
    currentCtx.prevHeight = height;
    currentCtx.prevWidth = width;
  }
}
const MIN_WIDTH_CANVAS = 15;
const Canvas = (0, import_react.memo)(
  (0, import_react.forwardRef)(
    ({ children, ...props }, ref) => {
      const { height, width } = props;
      const canvasRef = (0, import_react.useRef)(null);
      const { shapeRefs } = (0, import_react.useContext)(import_CanvasContext.CanvasContext);
      const contextRef = (0, import_react.useRef)({
        ctx: null,
        prevHeight: 0,
        prevWidth: 0
      });
      (0, import_react.useImperativeHandle)(
        ref,
        () => {
          return {
            customSetStyle: (key, value) => {
              canvasRef.current?.style.setProperty(key, value);
            }
          };
        },
        []
      );
      const isReady = (0, import_react.useDeferredValue)(width > MIN_WIDTH_CANVAS);
      (0, import_react.useEffect)(() => {
        if (!isReady || !canvasRef) {
          return;
        }
        ensureCanvasContext(contextRef.current, canvasRef, width, height);
        if (contextRef.current.ctx) {
          const shapes = [...shapeRefs.values()];
          requestAnimationFrame(() => {
            if (contextRef.current.ctx) {
              draw(canvasRef.current, contextRef.current.ctx, shapes);
            }
          });
        }
      });
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("canvas", { ref: canvasRef, children });
    }
  )
);
