// @dynatrace/strato-components/layouts/app-header/action-group/ActionGroup.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ActionGroup_exports = {};
__export(ActionGroup_exports, {
  ActionGroup: () => ActionGroup
});
module.exports = __toCommonJS(ActionGroup_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_clsx = __toESM(require("clsx"));
var import_react2 = require("react");
var import_react_is = require("react-is");
var import_ActionButton = require("./ActionButton.js");
var import_ActionGroup_css = require("./ActionGroup.sty.js");
var import_CollapsedActionGroup = require("./CollapsedActionGroup.js");
var import_constants = require("./constants.js");
var import_ActionGroupContext = require("./contexts/ActionGroupContext.js");
var import_Group = require("./Group.js");
var import_NavItem = require("./NavItem.js");
var import_useContainerSize = require("../../../core/hooks/useContainerSize.js");
var import_useMeasureChildren = require("../../../core/hooks/useMeasureChildren.js");
var import_useMergeRefs = require("../../../core/hooks/useMergeRefs.js");
var import_useRovingFocus = require("../../../core/hooks/useRovingFocus.js");
var import_merge_props = require("../../../core/utils/merge-props.js");
var import_Divider = require("../../divider/Divider.js");
var import_Flex = require("../../flex/Flex.js");
const COMPONENT_NAME = "ActionGroup";
function itemIsPinned(child) {
  return (0, import_react_is.isElement)(child) && typeof child.props === "object" && child.props !== null && "pin" in child.props && child.props.pin === true;
}
const ActionGroupRoot = (0, import_react2.forwardRef)(
  (props, forwardedRef) => {
    const {
      collapsible = true,
      collapseFrom = "start",
      children,
      rovingFocus = true,
      triggerIcon,
      density = "default",
      onCollapsedChange,
      onOverflowChange,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      ...restProps
    } = props;
    const _containerRef = (0, import_react2.useRef)(null);
    const { containerRef, measurements, childProps } = (0, import_useMeasureChildren.useMeasureChildren)();
    const measuredChildren = (0, import_react2.useRef)([]);
    const { width } = (0, import_useContainerSize.useContainerSize)(_containerRef);
    const containerWidth = Math.round(width);
    const menuItems = (0, import_react2.useRef)(/* @__PURE__ */ new Map());
    const pinnedItemsCount = (0, import_react2.useRef)(0);
    const [isCollapsed, setIsCollapsed] = (0, import_react2.useState)(false);
    const [isOverflowing, setIsOverflowing] = (0, import_react2.useState)(false);
    const totalChildrenWidth = (0, import_react2.useMemo)(() => {
      let childWidths = Object.values(measurements).reduce(
        (acc, value) => acc += value?.width || 0,
        0
      );
      childWidths += (childProps.length - 1) * import_constants.ACTION_GROUP_GAP;
      return childWidths;
    }, [childProps, measurements]);
    const slots = (0, import_react2.useMemo)(() => {
      const slots2 = { items: [], menuItems: [] };
      const childrenArray = import_react2.Children.toArray(children);
      if (collapsible && containerWidth > 0 && childProps.length > 0 && Object.keys(measurements).length === childProps.length) {
        const childKeys = childProps.map((props2) => props2["data-id"]);
        let childIndex = collapseFrom === "start" ? 0 : childProps.length - 1;
        menuItems.current.clear();
        let childWidths = totalChildrenWidth;
        while (childWidths > containerWidth && (collapseFrom === "end" && childIndex >= 0 || collapseFrom === "start" && childIndex < childProps.length)) {
          const childNode = measuredChildren.current[
            // The childProps array also includes props for the menu trigger, which is not included in the
            // measured children array. If collapsing from the end, the index needs to be shifted back by 1.
            childIndex + (isOverflowing && collapseFrom === "end" ? -1 : 0)
          ];
          if ((0, import_react_is.isElement)(childNode) && itemIsPinned(childNode)) {
          } else {
            let childKeyIndex = childIndex;
            if (isOverflowing) {
              const collapseFromOffset = collapseFrom === "start" ? 1 : -1;
              childKeyIndex = childKeyIndex + collapseFromOffset;
            }
            const key = childKeys[childKeyIndex];
            if (key && measurements[key] !== void 0) {
              menuItems.current.set(key, true);
              childWidths -= (measurements[key]?.width || 0) + import_constants.ACTION_GROUP_GAP;
            }
          }
          childIndex += collapseFrom === "start" ? 1 : -1;
        }
      }
      let groupCount = 0;
      let pinnedItems = 0;
      measuredChildren.current = [];
      for (const [index, child] of childrenArray.entries()) {
        if (!(0, import_react_is.isElement)(child)) {
          continue;
        }
        const childKey = index + groupCount;
        const childIndex = childKey + (isOverflowing && collapseFrom === "start" ? 1 : 0);
        const isItemOverflowing = menuItems.current.get(
          childProps[childIndex]?.["data-id"] || ""
        );
        const clonedChild = (0, import_react2.cloneElement)(
          child,
          (0, import_merge_props.mergeProps)(child.props ?? {}, childProps[childIndex], {
            key: childKey,
            className: (0, import_clsx.default)(props.className, {
              [import_ActionGroup_css.actionGroupHiddenItemCSS]: isItemOverflowing
            })
          })
        );
        if (child.type === import_Group.Group) {
          const isNextItemOverflowing = menuItems.current.get(
            childProps[childIndex + 2]?.["data-id"] || ""
          );
          measuredChildren.current.push(clonedChild);
          slots2.items.push(
            /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_react2.Fragment, { children: [
              clonedChild,
              index !== childrenArray.length - 1 && /* @__PURE__ */ (0, import_react.createElement)(
                import_Divider.Divider,
                {
                  orientation: "vertical",
                  flexItem: false,
                  style: {
                    height: 20
                  },
                  className: (0, import_clsx.default)({
                    [import_ActionGroup_css.actionGroupHiddenItemCSS]: !!isItemOverflowing || !!isNextItemOverflowing
                  }),
                  ...childProps[childIndex + 1],
                  key: childKey + 1
                }
              )
            ] }, `fragment-${childKey}`)
          );
          if (itemIsPinned(child)) {
            measuredChildren.current.push({ props: { pin: true } });
            pinnedItems += 1;
          } else {
            measuredChildren.current.push({});
          }
          groupCount += 1;
        } else {
          measuredChildren.current.push(clonedChild);
          slots2.items.push(clonedChild);
          if (itemIsPinned(child)) {
            pinnedItems += 1;
          }
        }
        if (isItemOverflowing) {
          slots2.menuItems.push(child);
        }
      }
      pinnedItemsCount.current = pinnedItems;
      setIsOverflowing(slots2.menuItems.length > 0);
      setIsCollapsed(
        slots2.menuItems.length + pinnedItemsCount.current === slots2.items.length
      );
      return slots2;
    }, [
      measurements,
      childProps,
      children,
      collapseFrom,
      collapsible,
      totalChildrenWidth,
      containerWidth
    ]);
    const { containerRef: rovingContainerRef } = (0, import_useRovingFocus.useRovingFocus)(
      {
        // FIXME: Follow-up for detecting the initial index already, when it is initially undefined based on
        // either tab-index attributes or the first focusable element. This will solve the problem for
        // everyone and will not overcomplicate the logic. ADES-2260
        initialIndex: void 0,
        direction: "horizontal",
        focusableChildrenSelector: ':scope a, :scope button, :scope [type="button"]'
      }
    );
    const rootRef = (0, import_useMergeRefs.useMergeRefs)([
      forwardedRef,
      ...rovingFocus ? [rovingContainerRef] : []
    ]);
    const collapsibleRootRef = (0, import_useMergeRefs.useMergeRefs)([
      rootRef,
      containerRef,
      _containerRef
    ]);
    (0, import_react2.useEffect)(() => {
      onOverflowChange?.(isOverflowing);
    }, [isOverflowing, onOverflowChange]);
    (0, import_react2.useEffect)(() => {
      onCollapsedChange?.(isCollapsed);
    }, [isCollapsed, onCollapsedChange]);
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Flex.Flex,
      {
        "data-actiongroup": true,
        alignItems: "center",
        "data-testid": dataTestId,
        ref: collapsible ? collapsibleRootRef : rootRef,
        gap: import_constants.ACTION_GROUP_GAP,
        className: consumerClassName,
        style: { minWidth: !collapsible ? 0 : void 0, ...consumerStyle },
        ...restProps,
        "data-dt-component": COMPONENT_NAME,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
          import_ActionGroupContext.ActionGroupContext.Provider,
          {
            value: {
              isCollapsed: false
            },
            children: [
              collapseFrom === "start" && slots?.menuItems.length > 0 && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_CollapsedActionGroup.CollapsedActionGroup,
                {
                  triggerIcon,
                  childProps: childProps[0],
                  density,
                  children: slots?.menuItems
                }
              ),
              slots?.items,
              collapseFrom === "end" && slots?.menuItems.length > 0 && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_CollapsedActionGroup.CollapsedActionGroup,
                {
                  triggerIcon,
                  childProps: childProps[childProps.length - 1],
                  density,
                  children: slots?.menuItems
                }
              )
            ]
          }
        )
      }
    );
  }
);
ActionGroupRoot.displayName = "ActionGroup";
const ActionGroup = Object.assign(ActionGroupRoot, {
  ActionButton: import_ActionButton.ActionButton,
  Group: import_Group.Group,
  NavItem: import_NavItem.NavItem
});
