// @dynatrace/strato-components/layouts/app-header/action-group/ActionButton.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ActionButton_exports = {};
__export(ActionButton_exports, {
  ActionButton: () => ActionButton
});
module.exports = __toCommonJS(ActionButton_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var styles = __toESM(require("./ActionButton.sty.js"));
var import_useActionGroupContext = require("./hooks/useActionGroupContext.js");
var import_Button = require("../../../buttons/button/Button.js");
var import_focusRingStatic_css = require("../../../core/styles/focusRingStatic.sty.js");
var import_Menu = require("../../../navigation/menu/Menu.js");
var import_textStyle_css = require("../../../styles/textStyle.sty.js");
const ActionButton = (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const {
      showLabel = true,
      isSelected,
      children,
      prefixIcon,
      disabled,
      as,
      pin,
      onClick,
      density = "default",
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      ...remainingProps
    } = props;
    const { isCollapsed } = (0, import_useActionGroupContext.useActionGroupContext)();
    const showChildren = !prefixIcon || showLabel || isCollapsed;
    const sharedProps = {
      disabled,
      className: (0, import_clsx.default)(
        consumerClassName,
        styles.navItem({
          density
        }),
        styles.navItemSelectionStyle,
        {
          [styles.disabled]: disabled,
          active: isSelected
        }
      ),
      "data-testid": dataTestId
    };
    if (isCollapsed) {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
        import_Menu.Menu.Item,
        {
          ref: forwardedRef,
          as,
          onSelect: onClick ?? (() => null),
          ...sharedProps,
          style: consumerStyle,
          ...remainingProps,
          children: [
            prefixIcon && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Prefix, { children: prefixIcon }),
            showChildren && children
          ]
        }
      );
    }
    const sharedActionButtonProps = {
      ...sharedProps,
      onClick,
      className: (0, import_clsx.default)(
        (0, import_focusRingStatic_css.focusRingStatic)({ disabled }),
        (0, import_textStyle_css.textStyle)({ textStyle: "base-emphasized" }),
        sharedProps.className
      )
    };
    if (as) {
      const Component = as;
      return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
        Component,
        {
          ref: forwardedRef,
          ...sharedActionButtonProps,
          style: consumerStyle,
          ...remainingProps,
          children: [
            prefixIcon,
            showChildren && children
          ]
        }
      );
    }
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      import_Button.Button,
      {
        ref: forwardedRef,
        ...sharedActionButtonProps,
        style: {
          display: "flex",
          minWidth: 0,
          flexShrink: 0,
          ...consumerStyle
        },
        ...remainingProps,
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: prefixIcon }),
          showChildren && children
        ]
      }
    );
  }
);
ActionButton.displayName = "ActionGroup.Button";
