// @dynatrace/strato-components/layouts/app-header/action-group/CollapsedActionGroup.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CollapsedActionGroup_exports = {};
__export(CollapsedActionGroup_exports, {
  CollapsedActionGroup: () => CollapsedActionGroup
});
module.exports = __toCommonJS(CollapsedActionGroup_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_ActionGroupContext = require("./contexts/ActionGroupContext.js");
var import_Button = require("../../../buttons/button/Button.js");
var import_Menu = require("../../../navigation/menu/Menu.js");
const CollapsedActionGroup = (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    children,
    density = "default",
    triggerIcon = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.DotMenuIcon, {}),
    childProps
  } = props;
  const intl = (0, import_react_intl.useIntl)();
  const [open, setOpen] = (0, import_react.useState)(false);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ActionGroupContext.ActionGroupContext.Provider, { value: { isCollapsed: true }, children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Menu.Menu, { open, onOpenChange: setOpen, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Trigger, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Button.Button,
      {
        variant: "default",
        size: density,
        ...childProps,
        "aria-label": intl.formatMessage({
          id: "L/D6X7Ee5PYWzl3i",
          defaultMessage: "Show more actions"
        }),
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: triggerIcon })
      }
    ) }),
    open && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Content, { ref: forwardedRef, children })
  ] }) });
});
