// @dynatrace/strato-components/layouts/app-header/action-group/Group.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Group_exports = {};
__export(Group_exports, {
  Group: () => Group
});
module.exports = __toCommonJS(Group_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_is = require("react-is");
var import_constants = require("./constants.js");
var import_useActionGroupContext = require("./hooks/useActionGroupContext.js");
var import_Menu = require("../../../navigation/menu/Menu.js");
var import_Flex = require("../../flex/Flex.js");
const Group = (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const {
      children,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      ...remainingProps
    } = props;
    const slots = {
      items: []
    };
    for (const child of import_react.Children.toArray(children)) {
      if ((0, import_react_is.isElement)(child)) {
        slots.items?.push(child);
      }
    }
    const { isCollapsed } = (0, import_useActionGroupContext.useActionGroupContext)();
    return isCollapsed ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Menu.Menu.Group,
      {
        ref: forwardedRef,
        className: consumerClassName,
        style: consumerStyle,
        "data-testid": dataTestId,
        children: slots.items
      }
    ) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Flex.Flex,
      {
        ref: forwardedRef,
        gap: import_constants.ACTION_GROUP_GAP,
        alignItems: "center",
        role: "group",
        className: consumerClassName,
        style: { ...consumerStyle, flexShrink: 0 },
        "data-testid": dataTestId,
        ...remainingProps,
        children: slots.items
      }
    );
  }
);
Group.displayName = "ActionGroup.Group";
