// @dynatrace/strato-components/esm/charts/histogram/components/overlays/utils/bins-to-tooltip.js
import { getFormattedValues } from "./get-formatted-values.js";
import { safeGet } from "../../../providers/HistogramDataProvider.js";
function toTooltipItems(binsMap, formatters, isSingle, units, points, closestPoint) {
  if (!closestPoint) {
    return { tooltipItems: [], closestTooltipItem: void 0 };
  }
  const binMetadata = safeGet(binsMap, closestPoint.bin);
  const [leftUnit] = units;
  const [leftFormatter, rightFormatter] = formatters;
  const formatter = binMetadata.unit === leftUnit ? leftFormatter : rightFormatter;
  const closestTooltipItem = {
    id: binMetadata.seriesId,
    name: binMetadata.seriesName,
    color: binMetadata.color,
    asyncValues: () => getFormattedValues(formatter, binMetadata),
    bin: closestPoint.bin
  };
  if (isSingle) {
    return {
      tooltipItems: [closestTooltipItem],
      closestTooltipItem
    };
  }
  if (!points) {
    return { tooltipItems: [], closestTooltipItem };
  }
  const tooltipItems = points.sort((a, b) => a.y - b.y).map((point) => {
    const binSharedMetadata = safeGet(binsMap, point.bin);
    return {
      id: binSharedMetadata.seriesId,
      name: binSharedMetadata.seriesName,
      color: binSharedMetadata.color,
      asyncValues: () => getFormattedValues(formatter, binSharedMetadata),
      bin: point.bin
    };
  });
  return { tooltipItems, closestTooltipItem };
}
export {
  toTooltipItems
};
