// @dynatrace/strato-components/forms/number-input/NumberInputBase.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var NumberInputBase_exports = {};
__export(NumberInputBase_exports, {
  NumberInputBase: () => NumberInputBase
});
module.exports = __toCommonJS(NumberInputBase_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_numberfield = require("@react-aria/numberfield");
var import_numberfield2 = require("@react-stately/numberfield");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useControllableState = require("../../core/hooks/useControllableState.js");
var import_useFormControlState = require("../../core/hooks/useFormControlState.js");
var import_useFocusRing = require("../../core/styles/useFocusRing.js");
var import_merge_props = require("../../core/utils/merge-props.js");
var import_Flex = require("../../layouts/flex/Flex.js");
var import_InputGroup_css = require("../../layouts/input-group/InputGroup.sty.js");
var styles = __toESM(require("../_base-input/BaseInput.sty.js"));
var import_FormFieldMessageBaseItem = require("../forms-core/components/form-field-messages/FormFieldMessageBaseItem.js");
var import_useSyncedFormControlProps = require("../forms-core/hooks/useSyncedFormControlProps.js");
var import_helpers = require("../forms-core/time/helpers.js");
var import_useAutofillEvent_css = require("../hooks/useAutofillEvent.sty.js");
var import_useAutofillEvent = require("../hooks/useAutofillEvent.js");
var import_useInputHint = require("../hooks/useInputHint.js");
const NumberInputBase = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const intl = (0, import_react_intl.useIntl)();
  const defaultPlaceholder = intl.formatMessage({
    defaultMessage: "Add number",
    id: "rrkUQoVNb46Jbd/8",
    description: "Default placeholder for the number input."
  });
  const {
    id: propId,
    defaultValue,
    value,
    name,
    onChange,
    disabled: propDisabled,
    readOnly: propReadOnly,
    required: propRequired,
    min: minValueNumberLike,
    max: maxValueNumberLike,
    step,
    placeholder = defaultPlaceholder,
    controlState,
    onKeyUp,
    onKeyDown,
    onBlur,
    onFocus,
    onFocusChange,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "aria-disabled": ariaDisabledProp,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const componentName = restProps["data-dt-component"];
  const { "aria-labelledby": ariaLabelledBy, ...remainingAriaProps } = (0, import_useAriaLabelingProps.useAriaLabelingProps)(restProps);
  const locale = (0, import_react.useMemo)(() => (0, import_helpers.getSafeRegionalFormat)(), []);
  const minValue = !isNaN(Number(minValueNumberLike)) ? Number(minValueNumberLike) : void 0;
  const maxValue = !isNaN(Number(maxValueNumberLike)) ? Number(maxValueNumberLike) : void 0;
  const numberInputValue = value === null || value === void 0 ? NaN : value;
  const {
    disabled,
    required,
    id: contextId,
    ariaDisabled,
    readOnly: isReadOnly
  } = (0, import_useSyncedFormControlProps.useSyncedFormControlProps)({
    propDisabled,
    propRequired,
    propId,
    propAriaDisabled: ariaDisabledProp,
    propReadOnly
  });
  const disabledOrAriaDisabled = disabled || ariaDisabled || false;
  const state = (0, import_numberfield2.useNumberFieldState)({
    value: numberInputValue,
    onChange,
    placeholder,
    minValue,
    maxValue,
    step,
    isDisabled: disabled,
    isReadOnly,
    isRequired: required,
    locale
  });
  const inputRef = (0, import_react.useRef)(null);
  const wrapperRef = (0, import_react.useRef)(null);
  (0, import_react.useImperativeHandle)(
    forwardedRef,
    () => ({
      element: wrapperRef.current,
      inputRef: inputRef.current,
      focus() {
        inputRef.current?.focus();
      }
    }),
    []
  );
  const { inputProps } = (0, import_numberfield.useNumberField)(
    {
      value: numberInputValue,
      onChange,
      id: contextId,
      placeholder,
      minValue,
      maxValue,
      step,
      isDisabled: disabled,
      isReadOnly,
      isRequired: required,
      onKeyUp,
      onKeyDown,
      onBlur,
      onFocus,
      onFocusChange,
      // we pass the aria-labelledBy only if there's no aria-label, because the useNumberField would add aria-labelledby
      // with 2 ids and append the input text to its aria-label otherwise, breaking the getByLabelText selectors in tests otherwise
      "aria-labelledby": remainingAriaProps["aria-label"] ? void 0 : ariaLabelledBy,
      ...remainingAriaProps
      // passing the aria props here prevents unnecessary react-aria warnings when aria-labelling props are already set
    },
    state,
    // eslint-disable-next-line @typescript-eslint/prefer-ts-expect-error, @typescript-eslint/ban-ts-comment
    // @ts-ignore React 19 - proper overlap for the ref is no longer there. It should be fine, but the Numberfield will be reworked anyways
    inputRef
  );
  let inputDefaultValue;
  if (defaultValue !== null && defaultValue !== void 0) {
    inputDefaultValue = defaultValue.toString();
  } else {
    inputDefaultValue = value === void 0 && !onChange ? "" : void 0;
  }
  const {
    value: inputPropsValue,
    type,
    id,
    readOnly = false,
    role,
    onKeyDown: onKeyDownReactAria,
    ...remainingInputProps
  } = inputProps;
  const [inputValue, setInputValue] = (0, import_useControllableState.useControllableState)({
    defaultValue: inputDefaultValue,
    value: inputPropsValue?.toString()
  });
  const [hasError, hint] = (0, import_useFormControlState.useFormControlState)(controlState);
  const { inputProps: inputHintAriaProps, hintProps } = (0, import_useInputHint.useInputHint)({
    hint,
    hasError
  });
  const { isAutoFilled, onAnimationStart } = (0, import_useAutofillEvent.useAutofillEvent)();
  const handleInputChange = (event) => {
    remainingInputProps?.onChange?.(event);
    setInputValue(event.target.value);
  };
  const { focusProps, focusClassName } = (0, import_useFocusRing.useFocusRing)({
    color: hasError && !(disabled || ariaDisabled) ? "critical" : "neutral",
    ignoreModality: true,
    disabled
  });
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_Flex.Flex,
    {
      gap: 4,
      flexDirection: "column",
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ref: wrapperRef,
      className: consumerClassName,
      style: consumerStyle,
      ...behavioralTrackingProps,
      "data-dt-component": componentName,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_Flex.Flex,
          {
            gap: 2,
            alignItems: "center",
            className: (0, import_clsx.default)(styles.wrapper, styles.variant["default"], {
              [styles.hasError]: hasError,
              [styles.disabled]: disabledOrAriaDisabled,
              [styles.readOnly]: readOnly
            }),
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              "div",
              {
                className: (0, import_clsx.default)(
                  styles.container,
                  import_InputGroup_css.inputGroupClassName,
                  focusClassName,
                  {
                    [styles.isAutoFilled]: isAutoFilled
                  }
                ),
                children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  "input",
                  {
                    id,
                    ref: inputRef,
                    type: "text",
                    role,
                    onChange: handleInputChange,
                    onAnimationStart,
                    disabled,
                    className: (0, import_clsx.default)(styles.input, import_useAutofillEvent_css.autofillCSS),
                    readOnly,
                    ...(0, import_merge_props.mergeProps)(
                      remainingInputProps,
                      restProps,
                      focusProps,
                      inputHintAriaProps
                    ),
                    name,
                    required,
                    value: inputValue,
                    "aria-disabled": ariaDisabled,
                    onKeyDown: (e) => {
                      if (e.key === "Home" && minValue === void 0 || e.key === "End" && maxValue === void 0) {
                        onKeyDown?.(e);
                        return;
                      }
                      onKeyDownReactAria?.(e);
                    }
                  }
                )
              }
            )
          }
        ),
        hint && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_FormFieldMessageBaseItem.InternalBaseItem,
          {
            variant: hasError ? "error" : "hint",
            disabled: disabled || ariaDisabled,
            ...hintProps,
            children: hint
          }
        )
      ]
    }
  );
});
NumberInputBase.displayName = "NumberInputBase";
