// @dynatrace/strato-components/filters/timeframe-selector/TimeframeSelector.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TimeframeSelector_exports = {};
__export(TimeframeSelector_exports, {
  TimeframeSelector: () => TimeframeSelector
});
module.exports = __toCommonJS(TimeframeSelector_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_PresetItem = require("./components/presets/PresetItem.js");
var import_Presets = require("./components/presets/Presets.js");
var import_CustomTrigger = require("./components/trigger/CustomTrigger.js");
var import_DisplayValue = require("./components/trigger/DisplayValue.js");
var import_Trigger = require("./components/trigger/Trigger.js");
var import_timeframeSelectorMessages = require("./constants/timeframeSelectorMessages.js");
var import_TimeframeSelectorRoot = require("./TimeframeSelectorRoot.js");
var import_useCallbackRef = require("../../core/hooks/useCallbackRef.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_timeframeSelector = require("../../core/time/timeframeStore/actions/timeframeSelector.js");
var import_createTimeframeStore = require("../../core/time/timeframeStore/createTimeframeStore.js");
var import_timeframeStoreProvider = require("../../core/time/timeframeStore/timeframeStoreProvider.js");
var import_useDateTimeMessages = require("../../core/time/useDateTimeMessages.js");
var import_useMinMax = require("../../core/time/useMinMax.js");
var import_validation = require("../../core/time/validation.js");
var import_helpers = require("../../forms/calendar/utils/helpers.js");
var import_useFormFieldMessagesContext = require("../../forms/forms-core/hooks/useFormFieldMessagesContext.js");
var import_useSyncedFormControlProps = require("../../forms/forms-core/hooks/useSyncedFormControlProps.js");
var import_date_time = require("../../forms/forms-core/time/date-time.js");
var import_timeParser = require("../../forms/forms-core/time/timeParser.js");
const STEPPER_DEBOUNCE_MS = 300;
function useSyncPropsAndState(store, memoedValue, isControlled, precision, formatString, required, formFieldMessagesDispatch, timeframeSelectorRef, debouncedOnChangeRef) {
  (0, import_react.useEffect)(
    () => {
      isControlled && store.setState((state) => {
        debouncedOnChangeRef?.current?.cancel();
        const error = (0, import_validation.validate)(memoedValue ?? null, state.min, state.max, {
          formMessageDispatch: formFieldMessagesDispatch,
          showError: false,
          checkOrder: true,
          ref: timeframeSelectorRef,
          expectedFormat: formatString,
          valueMode: "timeframeSelector",
          required,
          errorMessages: import_timeframeSelectorMessages.TIMEFRAMESELECTOR_ERROR_MESSAGES,
          formerError: state.timeframeSelectorError
        });
        const value = (0, import_timeParser.removePrecisionFromTimeFrame)(
          memoedValue,
          state.precision
        );
        const timeOffset = (0, import_helpers.getTimeOffsetFromInitialValue)(
          value,
          "range",
          state.min,
          state.max,
          precision
        );
        return {
          ...state,
          timeframeSelectorTimeframe: value,
          calendarValue: value,
          dateTimeRangePickerTimeframe: value,
          timeframeSelectorError: error,
          calendarError: error,
          dateTimeRangePickerError: error,
          timeOffset,
          required
        };
      });
    },
    // timeframeSelectorRef and debouncedOnChangeRef are stable refs — no need to include in deps
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [
      formFieldMessagesDispatch,
      formatString,
      isControlled,
      memoedValue,
      precision,
      required,
      store
    ]
  );
  (0, import_react.useEffect)(() => {
    store.setState((state) => {
      const preciseValue = isControlled ? memoedValue : state.timeframeSelectorTimeframe;
      const value = (0, import_timeParser.removePrecisionFromTimeFrame)(preciseValue, precision);
      return {
        ...state,
        precision,
        formatString,
        timeframeSelectorTimeframe: value,
        calendarValue: value,
        dateTimeRangePickerTimeframe: value
      };
    });
  }, [precision, formatString, store, isControlled, memoedValue]);
}
const _TimeframeSelector = (0, import_react.forwardRef)(
  ({
    value: propValue,
    defaultValue: propDefaultValue,
    onChange,
    min: propMin = import_date_time.GRAIL_MIN_DATE,
    max: propMax = import_date_time.GRAIL_MAX_DATE,
    disabled: propDisabled,
    required: propRequired,
    readOnly: propReadOnly,
    id: propId,
    "aria-disabled": propAriaDisabled,
    precision = "minutes",
    ...remainingProps
  }, forwardedRef) => {
    const [isControlled] = (0, import_react.useState)(propValue !== void 0);
    const onChangeProp = (0, import_useCallbackRef.useCallbackRef)(onChange);
    const memoedValue = (0, import_react.useMemo)(
      () => (0, import_timeParser.mapToPartialTimeframe)(propValue, precision),
      [precision, propValue]
    );
    const memoedDefaultValue = (0, import_react.useMemo)(
      () => (0, import_timeParser.mapToPartialTimeframe)(propDefaultValue, precision),
      /** We're ignoring updates to defaultValue as it only considers the initial state. */
      // eslint-disable-next-line react-hooks/exhaustive-deps
      []
    );
    const { min, max } = (0, import_useMinMax.useMinMax)(propMin, propMax);
    const calendarRef = (0, import_react.useRef)(null);
    const dateTimeFromRef = (0, import_react.useRef)(null);
    const dateTimeRangePickerRef = (0, import_react.useRef)(null);
    const timeframeSelectorRef = (0, import_react.useRef)(null);
    const { formatString } = (0, import_useDateTimeMessages.useDateTimeMessages)({ precision });
    const { dispatch: formFieldMessagesDispatch } = (0, import_useFormFieldMessagesContext.useFormFieldMessagesContext)();
    const { disabled, required, id, ariaDisabled, readOnly } = (0, import_useSyncedFormControlProps.useSyncedFormControlProps)({
      propDisabled,
      propRequired,
      propId,
      propAriaDisabled,
      propReadOnly
    });
    const debouncedOnChangeRef = (0, import_react.useRef)(
      (0, import_lodash_es.debounce)((timeframe) => {
        onChangeProp?.(timeframe);
      }, STEPPER_DEBOUNCE_MS)
    );
    (0, import_react.useEffect)(() => {
      const debouncedOnChange = debouncedOnChangeRef.current;
      return () => debouncedOnChange.cancel();
    }, []);
    const [initialState] = (0, import_react.useState)(() => {
      const value = isControlled ? memoedValue : memoedDefaultValue;
      const error = (0, import_validation.validate)(value, min, max, {
        formMessageDispatch: formFieldMessagesDispatch,
        precision,
        showError: false,
        checkOrder: true,
        ref: timeframeSelectorRef,
        expectedFormat: formatString,
        valueMode: "timeframeSelector",
        required,
        errorMessages: import_timeframeSelectorMessages.TIMEFRAMESELECTOR_ERROR_MESSAGES
      });
      const timeOffset = (0, import_helpers.getTimeOffsetFromInitialValue)(
        value,
        "range",
        min,
        max,
        precision
      );
      return {
        calendarValue: value,
        dateTimeRangePickerTimeframe: value,
        timeframeSelectorTimeframe: value,
        referenceTimeframe: null,
        min,
        max,
        precision,
        isFullDay: false,
        calendarRef,
        dateTimeFromRef,
        dateTimeRangePickerRef,
        timeframeSelectorRef,
        required,
        timeframeSelectorError: error,
        calendarError: error,
        dateTimeRangePickerError: error,
        previousTimeframe: null,
        timeOffset,
        formatString
      };
    });
    const [store] = (0, import_react.useState)(
      () => (0, import_createTimeframeStore.createTimeframeStore)(
        initialState,
        (0, import_timeframeSelector.getTimeframeSelectorActions)(
          isControlled,
          timeframeSelectorRef,
          formFieldMessagesDispatch,
          onChangeProp,
          debouncedOnChangeRef,
          import_timeframeSelectorMessages.TIMEFRAMESELECTOR_ERROR_MESSAGES
        )
      )
    );
    useSyncPropsAndState(
      store,
      memoedValue,
      isControlled,
      precision,
      formatString,
      required,
      formFieldMessagesDispatch,
      timeframeSelectorRef,
      debouncedOnChangeRef
    );
    (0, import_useMinMax.useMinMaxStoreSync)(
      store,
      min,
      max,
      (0, import_react.useCallback)(
        (state, min2, max2) => {
          const error = (0, import_validation.validate)(state.timeframeSelectorTimeframe, min2, max2, {
            formMessageDispatch: formFieldMessagesDispatch,
            formerError: state.timeframeSelectorError,
            precision: state.precision,
            showError: false,
            checkOrder: true,
            ref: timeframeSelectorRef,
            expectedFormat: state.formatString,
            valueMode: "timeframeSelector",
            required: state.required,
            errorMessages: import_timeframeSelectorMessages.TIMEFRAMESELECTOR_ERROR_MESSAGES
          });
          return {
            timeframeSelectorError: error,
            calendarError: error,
            dateTimeRangePickerError: error
          };
        },
        [formFieldMessagesDispatch]
      )
    );
    const mergedRef = (0, import_useMergeRefs.useMergeRefs)([
      timeframeSelectorRef,
      forwardedRef
    ]);
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_timeframeStoreProvider.TimeframeSelectorStoreProvider, { store, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_TimeframeSelectorRoot.TimeframeSelectorRoot,
      {
        ...remainingProps,
        id,
        ref: mergedRef,
        "aria-disabled": ariaDisabled,
        disabled,
        readOnly
      }
    ) });
  }
);
_TimeframeSelector.displayName = "TimeframeSelector";
const TimeframeSelector = Object.assign(_TimeframeSelector, {
  Trigger: import_Trigger.Trigger,
  DisplayValue: import_DisplayValue.DisplayValue,
  CustomTrigger: import_CustomTrigger.CustomTrigger,
  Presets: import_Presets.Presets,
  PresetItem: import_PresetItem.PresetItem
});
