// @dynatrace/strato-components/filters/timeframe-selector/TimeframeSelectorRoot.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TimeframeSelectorRoot_exports = {};
__export(TimeframeSelectorRoot_exports, {
  TimeframeSelectorRoot: () => TimeframeSelectorRoot
});
module.exports = __toCommonJS(TimeframeSelectorRoot_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_TimeframeSelectorOverlay = require("./components/overlay/TimeframeSelectorOverlay.js");
var import_PresetItem = require("./components/presets/PresetItem.js");
var import_Presets = require("./components/presets/Presets.js");
var import_StepperButton = require("./components/stepper/StepperButton.js");
var import_TimeframeSelectorTooltip = require("./components/TimeframeSelectorTooltip.js");
var import_CustomTrigger = require("./components/trigger/CustomTrigger.js");
var import_Trigger = require("./components/trigger/Trigger.js");
var import_timeframeSelectorMessages = require("./constants/timeframeSelectorMessages.js");
var import_TimeframeSelectorContext = require("./contexts/TimeframeSelectorContext.js");
var styles = __toESM(require("./TimeframeSelector.sty.js"));
var import_getTimeframeDisplayValueProps = require("./utils/getTimeframeDisplayValueProps.js");
var import_presets_helpers = require("./utils/presets-helpers.js");
var import_TimeframePresetsClient = require("./utils/TimeframePresetsClient.js");
var import_VisuallyHidden = require("../../core/components/visually-hidden/VisuallyHidden.js");
var import_useFormControlState = require("../../core/hooks/useFormControlState.js");
var import_useId = require("../../core/hooks/useId.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_useTimeframeStore = require("../../core/time/timeframeStore/useTimeframeStore.js");
var import_validation = require("../../core/time/validation.js");
var import_is_slot = require("../../core/utils/is-slot.js");
var import_logging = require("../../core/utils/logging.js");
var import_FormFieldMessageBaseItem = require("../../forms/forms-core/components/form-field-messages/FormFieldMessageBaseItem.js");
var import_useFormFieldMessagesContext = require("../../forms/forms-core/hooks/useFormFieldMessagesContext.js");
var import_useFormFieldValidation = require("../../forms/forms-core/hooks/useFormFieldValidation.js");
var import_useMinMaxLogger = require("../../forms/forms-core/hooks/useMinMaxLogger.js");
var import_formfield_messages_state_reducer = require("../../forms/forms-core/state/formfield-messages-state-reducer.js");
var import_timeParser = require("../../forms/forms-core/time/timeParser.js");
var import_Flex = require("../../layouts/flex/Flex.js");
var import_InputGroup = require("../../layouts/input-group/InputGroup.js");
var import_useOverlayWithTrigger = require("../../overlays/hooks/useOverlayWithTrigger.js");
const COMPONENT_NAME = "TimeframeSelector";
function getItemSlots(children, precision) {
  const presetItemsDisplayValuesMap = /* @__PURE__ */ new Map();
  const slots = {
    trigger: null,
    customTrigger: null,
    presets: null,
    presetItemsDisplayValuesMap
  };
  for (const child of import_react.Children.toArray(children)) {
    if ((0, import_is_slot.isSlot)(child, import_Presets.Presets)) {
      slots.presets = child;
      for (const childItem of import_react.Children.toArray(child.props.children)) {
        if ((0, import_is_slot.isSlot)(childItem, import_PresetItem.PresetItem)) {
          slots.presetItemsDisplayValuesMap.set(
            `${(0, import_timeParser.removePrecision)(
              childItem.props.value.from,
              precision
            )} ${(0, import_timeParser.removePrecision)(childItem.props.value.to, precision)}`,
            childItem.props.children
          );
        }
      }
    } else if ((0, import_CustomTrigger.isTimeframeSelectorCustomTriggerElement)(child)) {
      slots.customTrigger = child;
    } else if ((0, import_Trigger.isTimeframeSelectorTriggerElement)(child)) {
      slots.trigger = child;
    } else {
      (0, import_logging.devLog)("The used child is not supported by the TimeframeSelector.");
    }
  }
  return slots;
}
const TimeframeSelectorRoot = (0, import_react.forwardRef)((props, forwardedRef) => {
  const intl = (0, import_react_intl.useIntl)();
  const {
    disabled,
    value: propValue,
    defaultValue,
    /* eslint-disable @typescript-eslint/no-deprecated */
    clearable = false,
    readOnly = false,
    controlState,
    children,
    onChange,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    name,
    id,
    stepper = true,
    "aria-disabled": ariaDisabled,
    "aria-label": ariaLabelProp,
    "aria-labelledby": ariaLabelledByProp,
    "aria-describedby": ariaDescribedByProp,
    "aria-details": ariaDetailsProp,
    ...remainingProps
  } = props;
  const value = (0, import_useTimeframeStore.useTimeframeSelectorStore)(
    (state) => state.timeframeSelectorTimeframe
  );
  const precision = (0, import_useTimeframeStore.useTimeframeSelectorStore)((state) => state.precision);
  const min = (0, import_useTimeframeStore.useTimeframeSelectorStore)((state) => state.min);
  const max = (0, import_useTimeframeStore.useTimeframeSelectorStore)((state) => state.max);
  const required = (0, import_useTimeframeStore.useTimeframeSelectorStore)((state) => state.required);
  const formatString = (0, import_useTimeframeStore.useTimeframeSelectorStore)((state) => state.formatString);
  const timeframeSelectorRef = (0, import_useTimeframeStore.useTimeframeSelectorStore)(
    (state) => state.timeframeSelectorRef
  );
  const setDateTimeRangePickerTimeframe = (0, import_useTimeframeStore.useTimeframeSelectorStore)(
    (state) => state.setDateTimeRangePickerTimeframe
  );
  const placeholder = intl.formatMessage({
    defaultMessage: "Select timeframe",
    id: "W+rTy8UuFKDho3mc",
    description: "Placeholder shown for the timeframe selector."
  });
  (0, import_useMinMaxLogger.useMinMaxLogger)(min, max, "TimeframeSelector");
  const [formControlStateError, hint] = (0, import_useFormControlState.useFormControlState)(controlState);
  const { state: formFieldMessagesState, dispatch: formFieldMessagesDispatch } = (0, import_useFormFieldMessagesContext.useFormFieldMessagesContext)();
  const onOpenChange = (0, import_react.useCallback)(
    (isOpen) => {
      if (!isOpen) {
        formFieldMessagesDispatch?.({
          type: import_formfield_messages_state_reducer.FormFieldMessagesAction.SETISTOUCHED,
          payload: { isTouched: true }
        });
        setDateTimeRangePickerTimeframe(value);
      }
    },
    [formFieldMessagesDispatch, setDateTimeRangePickerTimeframe, value]
  );
  const {
    targetRef,
    overlayRef,
    triggerRef,
    overlayProps,
    overlayState,
    overlayTriggerProps
  } = (0, import_useOverlayWithTrigger.useOverlayWithTrigger)({
    offset: 4,
    placement: "bottom-start",
    placementFallbacks: ["bottom-end", "top-start", "top-end"],
    heightStrategy: "maxViewport",
    disabled,
    onOpenChange
  });
  (0, import_react.useEffect)(() => {
    if (import_TimeframePresetsClient.timeframeQueryClient.getQueryData(import_TimeframePresetsClient.TIMEFRAME_PRESETS_QUERY_KEY)) {
      return;
    }
    import_TimeframePresetsClient.timeframeQueryClient.prefetchQuery({
      queryKey: import_TimeframePresetsClient.TIMEFRAME_PRESETS_QUERY_KEY,
      queryFn: import_presets_helpers.getTimeframePresets
    }).catch(() => void 0);
  }, []);
  const details = (0, import_react.useMemo)(() => {
    const { from, to } = (0, import_timeParser.parseValue)(value);
    return from && to ? { from, to } : null;
  }, [value]);
  const { customTrigger, trigger, presets, presetItemsDisplayValuesMap } = (0, import_react.useMemo)(() => getItemSlots(children, precision), [children, precision]);
  const presetItemName = presetItemsDisplayValuesMap.get(
    `${details?.from?.value} ${details?.to?.value}`
  );
  const {
    displayValue,
    isInvalid: displayValueError,
    hint: displayValueHint
  } = (0, import_getTimeframeDisplayValueProps.getTimeframeDisplayValueProps)(
    value,
    intl,
    placeholder,
    precision,
    presetItemName,
    required,
    min?.value,
    max?.value
  );
  const triggerErrorId = (0, import_useId.useId)();
  const inputId = (0, import_useId.useId)();
  const imperativeTriggerRef = (0, import_react.useRef)(null);
  const containerRef = (0, import_react.useRef)(null);
  const fromInputRef = (0, import_react.useRef)(null);
  const toInputRef = (0, import_react.useRef)(null);
  const ariaLabelingProps = {
    "aria-label": ariaLabelProp,
    "aria-labelledby": ariaLabelledByProp,
    "aria-describedby": ariaDescribedByProp,
    "aria-details": ariaDetailsProp
  };
  const ariaFormProps = (0, import_useFormFieldValidation.useFormFieldValidationAriaProps)(ariaLabelingProps);
  (0, import_react.useImperativeHandle)(
    forwardedRef,
    () => ({
      element: containerRef.current,
      triggerRef: imperativeTriggerRef.current,
      inputRef: { from: fromInputRef.current, to: toInputRef.current },
      open: overlayState.open,
      close: overlayState.close,
      focus() {
        imperativeTriggerRef.current?.focus?.();
      },
      validate() {
        formFieldMessagesDispatch?.({
          type: import_formfield_messages_state_reducer.FormFieldMessagesAction.SETISTOUCHED,
          payload: { isTouched: true }
        });
        (0, import_validation.validate)(value, min, max, {
          formMessageDispatch: formFieldMessagesDispatch,
          precision,
          showError: false,
          checkOrder: true,
          ref: timeframeSelectorRef,
          expectedFormat: formatString,
          valueMode: "timeframeSelector",
          required,
          errorMessages: import_timeframeSelectorMessages.TIMEFRAMESELECTOR_ERROR_MESSAGES
        });
      }
    }),
    [formFieldMessagesDispatch, overlayState.close, overlayState.open]
  );
  (0, import_react.useEffect)(() => {
    if (overlayState.isOpen) {
      return;
    }
    if (!displayValueError && triggerErrorId) {
      formFieldMessagesDispatch?.({
        type: import_formfield_messages_state_reducer.FormFieldMessagesAction.REMOVEMESSAGE,
        payload: {
          id: triggerErrorId,
          messageType: "formcontrol",
          controlId: inputId
        }
      });
    }
  }, [
    displayValueError,
    displayValueHint,
    overlayState.isOpen,
    triggerErrorId,
    formFieldMessagesDispatch,
    inputId
  ]);
  const triggerButtonRef = (0, import_useMergeRefs.useMergeRefs)([
    targetRef,
    triggerRef,
    imperativeTriggerRef
  ]);
  const handleOnBlur = (0, import_react.useCallback)(() => {
    if (!overlayState.isOpen) {
      fromInputRef.current?.reportValidity();
      formFieldMessagesDispatch?.({
        type: import_formfield_messages_state_reducer.FormFieldMessagesAction.SETISTOUCHED,
        payload: { isTouched: true }
      });
    }
  }, [overlayState.isOpen, formFieldMessagesDispatch]);
  const triggerProps = (0, import_react.useMemo)(
    () => ({
      disabled,
      onBlur: handleOnBlur,
      ref: triggerButtonRef,
      id,
      "aria-disabled": ariaDisabled,
      ...ariaFormProps,
      ...readOnly ? {} : overlayTriggerProps,
      "data-dt-component": COMPONENT_NAME
    }),
    [
      disabled,
      handleOnBlur,
      triggerButtonRef,
      id,
      ariaDisabled,
      ariaFormProps,
      readOnly,
      overlayTriggerProps
    ]
  );
  const timeframeSelectorContextValue = (0, import_react.useMemo)(
    () => ({
      details,
      displayValue,
      triggerProps,
      hint: displayValueHint,
      readOnly,
      showError: formControlStateError || formFieldMessagesState.showError,
      disabled
    }),
    [
      disabled,
      details,
      readOnly,
      displayValue,
      triggerProps,
      displayValueHint,
      formControlStateError,
      formFieldMessagesState.showError
    ]
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TimeframeSelectorContext.TimeframeSelectorContext.Provider, { value: timeframeSelectorContextValue, children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_Flex.Flex,
    {
      gap: 4,
      flexDirection: "column",
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ref: containerRef,
      className: (0, import_clsx.default)(consumerClassName, styles.root),
      style: consumerStyle,
      ...remainingProps,
      "data-dt-component": "TimeframeSelector",
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_VisuallyHidden.VisuallyHidden, { children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "input",
            {
              id: inputId,
              type: "text",
              ref: fromInputRef,
              name: name && `${name}-from`,
              "aria-hidden": true,
              readOnly,
              required,
              disabled,
              value: details?.from.value ?? "",
              onInvalid: (event) => {
                formFieldMessagesDispatch?.({
                  type: import_formfield_messages_state_reducer.FormFieldMessagesAction.SETISTOUCHED,
                  payload: { isTouched: true }
                });
                event.preventDefault();
              },
              onChange: () => {
              },
              tabIndex: -1,
              ...ariaFormProps
            }
          ),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "input",
            {
              type: "text",
              ref: toInputRef,
              name: name && `${name}-to`,
              "aria-hidden": true,
              readOnly,
              required,
              disabled,
              value: details?.to.value ?? "",
              onInvalid: (event) => event.preventDefault(),
              onChange: () => {
              },
              tabIndex: -1,
              ...ariaFormProps
            }
          )
        ] }),
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_InputGroup.InputGroup, { style: { gridAutoColumns: "1fr max-content max-content" }, children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_TimeframeSelectorTooltip.TimeframeSelectorTooltip,
            {
              displayValue,
              value,
              disabled,
              children: customTrigger ?? trigger ?? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Trigger.Trigger, {})
            }
          ),
          stepper && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_StepperButton.StepperButton,
              {
                disabled: disabled || readOnly,
                direction: "backward",
                "aria-disabled": ariaDisabled,
                onBlur: handleOnBlur
              }
            ),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_StepperButton.StepperButton,
              {
                disabled: disabled || readOnly,
                direction: "forward",
                "aria-disabled": ariaDisabled,
                onBlur: handleOnBlur
              }
            )
          ] })
        ] }),
        controlState && hint && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_FormFieldMessageBaseItem.InternalBaseItem,
          {
            variant: formControlStateError ? "error" : "hint",
            disabled: disabled || !!ariaDisabled,
            children: hint
          }
        ),
        overlayState.isOpen && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_TimeframeSelectorOverlay.TimeframeSelectorOverlay,
          {
            ref: overlayRef,
            overlayProps,
            overlayClose: overlayState.close,
            clearable,
            "data-dtrum-mask": dataDtrumMask,
            "data-dtrum-allow": dataDtrumAllow,
            children: presets
          }
        )
      ]
    }
  ) });
});
