// @dynatrace/strato-components/esm/charts/categorical-bar/components/interactions/datapoints/CategoricalBarChartDatapoints.js
import { jsx } from "react/jsx-runtime";
import { Canvas } from "../../../../core/components/canvas/Canvas.js";
import { CanvasContext } from "../../../../core/components/canvas/CanvasContext.js";
import { useGridPlotSize } from "../../../../core/components/chart-grid/hooks/useGridPlotSize.js";
import { OVERLAY_PADDING } from "../../../../core/components/data-point/constants.js";
import { DataPointCanvas } from "../../../../core/components/data-point/DataPointCanvas.js";
import { useCategoricalBarChartBaseConfig } from "../../../hooks/useCategoricalBarChartBaseConfig.js";
const CategoricalBarChartDatapoints = (props) => {
  const { width, height } = useGridPlotSize();
  const { layout } = useCategoricalBarChartBaseConfig();
  const { points, closestPoint } = props;
  return /* @__PURE__ */ jsx(
    CanvasContext.Provider,
    {
      value: {
        shapeRefs: /* @__PURE__ */ new Map()
      },
      children: /* @__PURE__ */ jsx(
        Canvas,
        {
          width: width + OVERLAY_PADDING * 2,
          height: height + OVERLAY_PADDING * 2,
          children: points && points.map((pointGroup, index) => /* @__PURE__ */ jsx(
            DataPointCanvas,
            {
              valueAxis: layout === "vertical" ? "y" : "x",
              points: pointGroup,
              closestPoint
            },
            `pointGroup-${index}`
          ))
        }
      )
    }
  );
};
CategoricalBarChartDatapoints["displayName"] = "CategoricalBarChartDatapoints";
export {
  CategoricalBarChartDatapoints
};
