// @dynatrace/strato-components/charts/core/components/annotations/hooks/useMarkerEventHandlers.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useMarkerEventHandlers_exports = {};
__export(useMarkerEventHandlers_exports, {
  useMarkerEventHandlers: () => useMarkerEventHandlers
});
module.exports = __toCommonJS(useMarkerEventHandlers_exports);
var import_lodash_es = require("lodash");
var import_selectors = require("./store/selectors.js");
var import_state_setters = require("./store/state-setters.js");
var import_useMarkersInteractions = require("./useMarkersInteractions.js");
var import_AnnotationsDataInitialState = require("../providers/AnnotationsDataInitialState.js");
var import_calculate_tooltip_position = require("../utils/calculate-tooltip-position.js");
const useMarkerEventHandlers = (annotationsContainerRef) => {
  const updateTooltipState = (0, import_state_setters.useSetStateTooltip)();
  const updateTooltipSelection = (0, import_state_setters.useSetStateSelection)();
  const { pinnedGroup } = (0, import_selectors.useAnnotationsSelectionState)();
  const onMarkersInteraction = (0, import_useMarkersInteractions.useMarkersInteractions)();
  const isTooltipPinned = !(0, import_lodash_es.isUndefined)(pinnedGroup);
  const handleMouseMove = (event) => {
    if (isTooltipPinned || !annotationsContainerRef?.current) {
      return;
    }
    const { clientX, clientY } = event;
    const absolutePosition = (0, import_calculate_tooltip_position.calculateTooltipPosition)(clientX, clientY);
    const annotationsContainerRect = annotationsContainerRef.current.getBoundingClientRect();
    updateTooltipState({
      position: absolutePosition,
      relativePosition: {
        x: absolutePosition.x - annotationsContainerRect.left,
        y: absolutePosition.y - annotationsContainerRect.top
      }
    });
  };
  const handleMouseEnter = (markerId, groupId, annotationsGroup) => (event) => {
    if (isTooltipPinned || !annotationsContainerRef?.current) {
      return;
    }
    const { clientX, clientY } = event;
    const absolutePosition = (0, import_calculate_tooltip_position.calculateTooltipPosition)(clientX, clientY);
    const annotationsContainerRect = annotationsContainerRef.current.getBoundingClientRect();
    updateTooltipState({
      position: absolutePosition,
      relativePosition: {
        x: absolutePosition.x - annotationsContainerRect.left,
        y: absolutePosition.y - annotationsContainerRect.top
      },
      data: annotationsGroup,
      visibility: true
    });
    const selectionState = {
      hoveredItem: markerId,
      hoveredGroup: groupId,
      data: annotationsGroup
    };
    updateTooltipSelection(selectionState);
    onMarkersInteraction?.(selectionState);
  };
  const handleClick = (groupId, annotationsGroup) => (event) => {
    event.stopPropagation();
    const { clientX, clientY } = event;
    const newSelection = {
      hoveredGroup: groupId,
      data: annotationsGroup,
      pinnedGroup: isTooltipPinned ? void 0 : groupId
      // if tooltip was already pinned, unpin
    };
    updateTooltipState({
      position: (0, import_calculate_tooltip_position.calculateTooltipPosition)(clientX, clientY),
      data: annotationsGroup,
      visibility: true,
      pinned: !isTooltipPinned
      // if tooltip was already pinned, show updated info but unpin
    });
    updateTooltipSelection(newSelection);
    onMarkersInteraction?.(newSelection);
  };
  const handleMouseLeave = () => {
    if (isTooltipPinned) {
      return;
    }
    updateTooltipState(import_AnnotationsDataInitialState.initialState.tooltip);
    updateTooltipSelection(import_AnnotationsDataInitialState.initialState.selection);
    onMarkersInteraction?.(import_AnnotationsDataInitialState.initialState.selection);
  };
  return {
    handleMouseEnter,
    handleMouseMove,
    handleMouseLeave,
    handleClick
  };
};
