// @dynatrace/strato-components/charts/core/components/tooltip/ChartTooltip/ChartTooltip.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ChartTooltip_exports = {};
__export(ChartTooltip_exports, {
  ChartTooltip: () => ChartTooltip,
  _ChartTooltip: () => _ChartTooltip
});
module.exports = __toCommonJS(ChartTooltip_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_dom = require("react-dom");
var import_ChartTooltip_css = require("./ChartTooltip.sty.js");
var import_useMergeRefs = require("../../../../../core/hooks/useMergeRefs.js");
var import_DataDtrumContext = require("../../../context/DataDtrumContext.js");
var import_is_touchable_device = require("../../../utils/is-touchable-device.js");
var import_ActionSelectionContext = require("../context/ActionSelectionContext.js");
var import_ChartTooltipMetadataContext = require("../context/ChartTooltipMetadataContext.js");
var import_useChartTooltip = require("../shared/hooks/useChartTooltip.js");
var import_ChartTooltip_config = require("../types/ChartTooltip.config.js");
const defaultSelectionValue = {
  selectedItemId: "",
  actionsMenuRef: void 0,
  updateSelectedItem: () => null
};
const _ChartTooltip = (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    children,
    pinned = false,
    position,
    updatePositionWhenPinned = false,
    actionSelectionValue = defaultSelectionValue,
    onMouseLeave,
    className,
    style
  } = props;
  const tooltipRef = (0, import_react.useRef)(null);
  const mergedRefs = (0, import_useMergeRefs.useMergeRefs)([forwardedRef, tooltipRef]);
  const { placement, updateTooltipPosition } = (0, import_useChartTooltip.useChartTooltip)(
    tooltipRef.current
  );
  const { dataDtrumMask, dataDtrumAllow } = (0, import_DataDtrumContext.useDataDtrumContext)();
  (0, import_react.useEffect)(() => {
    if (!pinned || updatePositionWhenPinned || (0, import_is_touchable_device.isTouchableDevice)()) {
      updateTooltipPosition(position);
    }
  }, [
    pinned,
    position.x,
    position.y,
    updatePositionWhenPinned,
    updateTooltipPosition
  ]);
  const visibility = tooltipRef.current ? "visible" : "hidden";
  const chartTooltipContent = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ActionSelectionContext.ActionSelectionContext.Provider, { value: actionSelectionValue, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ChartTooltipMetadataContext.ChartTooltipMetadataContext.Provider, { value: { tooltipRef }, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      className: (0, import_ChartTooltip_css.wrapper)({ isPinned: pinned }),
      "data-testid": "chart-tooltip",
      role: "tooltip",
      ref: mergedRefs,
      "data-placement": placement,
      onMouseLeave,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        "div",
        {
          className: (0, import_clsx.default)(import_ChartTooltip_css.chartTooltip, className),
          style: { ...style, visibility },
          children
        }
      )
    }
  ) }) });
  return (0, import_react_dom.createPortal)(chartTooltipContent, document.body);
});
_ChartTooltip.displayName = "ChartTooltip";
const ChartTooltip = Object.assign(
  _ChartTooltip,
  import_ChartTooltip_config.ChartTooltipSlots,
  import_ChartTooltip_config.ChartTooltipAtoms,
  import_ChartTooltip_config.ChartTooltipTemplate
);
