// @dynatrace/strato-components/forms/calendar/contexts/CalendarContext.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CalendarContext_exports = {};
__export(CalendarContext_exports, {
  CalendarContext: () => CalendarContext,
  initialState: () => initialState
});
module.exports = __toCommonJS(CalendarContext_exports);
var import_tz = require("@date-fns/tz");
var import_react = require("react");
var import_user_preferences = require("@dynatrace-sdk/user-preferences");
const today = import_tz.TZDateMini.tz((0, import_user_preferences.getTimezone)());
const initialState = {
  activeDate: today,
  currentView: "day",
  shouldUpdateFocusFromKeydown: null,
  today,
  type: "single"
};
const CalendarContext = (0, import_react.createContext)({
  state: initialState,
  dispatch: () => null,
  onChange: () => null,
  readOnly: false
});
