// @dynatrace/strato-components/forms/radio/RadioGroup.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var RadioGroup_exports = {};
__export(RadioGroup_exports, {
  RadioGroup: () => RadioGroup
});
module.exports = __toCommonJS(RadioGroup_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_RadioGroupContext = require("./RadioGroupContext.js");
var styles = __toESM(require("./styles.sty.js"));
var import_useControllableState = require("../../core/hooks/useControllableState.js");
var import_useFormControlState = require("../../core/hooks/useFormControlState.js");
var import_is_node_environment = require("../../core/utils/is-node-environment.js");
var import_Flex = require("../../layouts/flex/Flex.js");
var import_FormFieldMessageContextProvider = require("../forms-core/components/form-field/FormFieldMessageContextProvider.js");
var import_FormFieldMessageBaseItem = require("../forms-core/components/form-field-messages/FormFieldMessageBaseItem.js");
var import_useFormFieldMessagesContext = require("../forms-core/hooks/useFormFieldMessagesContext.js");
var import_useSyncedFormControlProps = require("../forms-core/hooks/useSyncedFormControlProps.js");
var import_formfield_messages_state_reducer = require("../forms-core/state/formfield-messages-state-reducer.js");
var import_useInputHint = require("../hooks/useInputHint.js");
const RadioGroupRoot = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    children,
    id: propId,
    disabled: propDisabled,
    value,
    name,
    defaultValue,
    controlState,
    onChange,
    readOnly: propReadOnly,
    required: propRequired,
    "aria-disabled": propAriaDisabled,
    "aria-describedby": ariaDescribedby,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    ...remainingProps
  } = props;
  const {
    disabled,
    required,
    id,
    ariaDisabled,
    readOnly: _readOnly
  } = (0, import_useSyncedFormControlProps.useSyncedFormControlProps)({
    propDisabled,
    propRequired,
    propId,
    propAriaDisabled,
    propReadOnly
  });
  const readOnly = !!_readOnly;
  const { state: messagesContextState, dispatch: messagesContextDispatch } = (0, import_useFormFieldMessagesContext.useFormFieldMessagesContext)();
  const [controlStateHasError, hint] = (0, import_useFormControlState.useFormControlState)(controlState);
  const hasError = controlState ? controlStateHasError : messagesContextState.showError;
  const { inputProps, hintProps } = (0, import_useInputHint.useInputHint)({
    ariaDescribedby,
    hint,
    hasError
  });
  const [controllableState, setControllableState] = (0, import_useControllableState.useControllableState)({
    defaultValue,
    value,
    onChange
  });
  const wrapperRef = (0, import_react.useRef)(null);
  (0, import_react.useImperativeHandle)(forwardedRef, () => {
    return {
      element: wrapperRef.current,
      focus() {
        const checkedInput = wrapperRef.current?.querySelector(
          "input:checked"
        );
        (checkedInput ?? wrapperRef.current?.querySelector("input"))?.focus();
      },
      validate() {
        if (controlState !== void 0) {
          return;
        }
        const inputs = wrapperRef.current?.querySelectorAll("input");
        inputs?.forEach((input) => {
          input?.checkValidity();
        });
        messagesContextDispatch?.({
          type: import_formfield_messages_state_reducer.FormFieldMessagesAction.SETISTOUCHED,
          payload: { isTouched: true }
        });
      }
    };
  });
  if ((0, import_is_node_environment.isDev)() && name === void 0) {
    console.warn(
      'RadioGroup: The "name" prop is missing, which is necessary to link the radio buttons correctly.'
    );
  }
  const radioGroupContextValue = (0, import_react.useMemo)(
    () => ({
      name,
      isDisabled: disabled,
      ariaDisabled,
      hasError,
      value: controllableState,
      onChange: setControllableState,
      readOnly,
      required,
      controlState
    }),
    [
      name,
      disabled,
      ariaDisabled,
      hasError,
      controllableState,
      setControllableState,
      readOnly,
      required,
      controlState
    ]
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    "div",
    {
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "aria-orientation": "vertical",
      role: "radiogroup",
      "aria-live": "polite",
      "aria-disabled": propAriaDisabled,
      ref: wrapperRef,
      className: (0, import_clsx.default)(consumerClassName, styles.group, {
        [styles.readOnly]: readOnly,
        [styles.hasError]: hasError
      }),
      style: consumerStyle,
      id,
      ...inputProps,
      ...remainingProps,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_RadioGroupContext.RadioGroupContext.Provider, { value: radioGroupContextValue, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Flex.Flex, { flexDirection: "column", gap: 8, children }) }),
        hint && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_FormFieldMessageBaseItem.InternalBaseItem,
          {
            variant: hasError ? "error" : "hint",
            disabled: disabled || ariaDisabled,
            ...hintProps,
            children: hint
          }
        )
      ]
    }
  );
});
const RadioGroup = (0, import_FormFieldMessageContextProvider.withFormFieldMessageContextProvider)(RadioGroupRoot);
RadioGroup.displayName = "RadioGroup";
