// @dynatrace/strato-components/forms/radio/Radio.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Radio_exports = {};
__export(Radio_exports, {
  Radio: () => Radio
});
module.exports = __toCommonJS(Radio_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_RadioGroupContext = require("./RadioGroupContext.js");
var styles = __toESM(require("./styles.sty.js"));
var import_VisuallyHidden = require("../../core/components/visually-hidden/VisuallyHidden.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useId = require("../../core/hooks/useId.js");
var import_useFocusRing = require("../../core/styles/useFocusRing.js");
var import_is_node_environment = require("../../core/utils/is-node-environment.js");
var import_merge_props = require("../../core/utils/merge-props.js");
var import_field_css = require("../../styles/field.sty.js");
var import_Text = require("../../typography/text/Text.js");
var import_useFormFieldMessagesContext = require("../forms-core/hooks/useFormFieldMessagesContext.js");
var import_useFormFieldValidation = require("../forms-core/hooks/useFormFieldValidation.js");
var import_formfield_messages_state_reducer = require("../forms-core/state/formfield-messages-state-reducer.js");
const COMPONENT_NAME = "Radio";
const Radio = /* @__PURE__ */ (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const fallbackId = (0, import_useId.useId)();
    const {
      disabled: propDisabled,
      id = fallbackId,
      children,
      value,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "aria-label": ariaLabelProp,
      "aria-labelledby": ariaLabelledByProp,
      "aria-describedby": ariaDescribedByProp,
      "aria-details": ariaDetailsProp,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
    const { dispatch: messagesContextDispatch } = (0, import_useFormFieldMessagesContext.useFormFieldMessagesContext)();
    const radioGroupContext = (0, import_react.useContext)(import_RadioGroupContext.RadioGroupContext);
    if ((0, import_is_node_environment.isDev)() && !radioGroupContext) {
      console.warn(
        "<Radio /> can only be used as a child element of <RadioGroup />"
      );
    }
    const {
      isDisabled,
      ariaDisabled,
      controlState,
      hasError,
      readOnly,
      name: groupName,
      required: groupRequired,
      value: groupValue,
      onChange
    } = radioGroupContext || {};
    const radioLabelRef = (0, import_react.useRef)(null);
    const inputRef = (0, import_react.useRef)(null);
    (0, import_react.useImperativeHandle)(forwardedRef, () => {
      return {
        element: radioLabelRef.current,
        inputRef: inputRef.current,
        focus() {
          inputRef.current?.focus();
        },
        validate() {
          if (controlState !== void 0) {
            return;
          }
          inputRef.current?.checkValidity();
          messagesContextDispatch?.({
            type: import_formfield_messages_state_reducer.FormFieldMessagesAction.SETISTOUCHED,
            payload: { isTouched: true }
          });
        }
      };
    });
    const {
      onChange: validationOnChange,
      onBlur: validationOnBlur,
      onInvalid: validationOnInvalid,
      ariaProps
    } = (0, import_useFormFieldValidation.useFormFieldValidation)(
      inputRef,
      { groupRequired },
      {
        "aria-label": ariaLabelProp,
        "aria-labelledby": ariaLabelledByProp,
        "aria-describedby": ariaDescribedByProp,
        "aria-details": ariaDetailsProp
      }
    );
    (0, import_react.useEffect)(() => {
      messagesContextDispatch?.({
        type: import_formfield_messages_state_reducer.FormFieldMessagesAction.ADDFORMCONTROLID,
        payload: { controlId: id }
      });
      return () => messagesContextDispatch?.({
        type: import_formfield_messages_state_reducer.FormFieldMessagesAction.REMOVEFORMCONTROLID,
        payload: { controlId: id }
      });
    }, [id]);
    const disabled = propDisabled ?? isDisabled;
    const disabledOrAriaDisabled = disabled || ariaDisabled || false;
    const checked = value === groupValue;
    function handleChange(e) {
      if (readOnly) {
        return;
      }
      onChange?.(e.target.value);
      !controlState && validationOnChange?.();
    }
    let variant = "default";
    let color = "neutral";
    if (checked) {
      variant = "accent";
    } else if (hasError) {
      variant = "emphasized";
    }
    if (hasError) {
      color = "critical";
    } else if (!readOnly && checked) {
      color = "primary";
    }
    const { focusProps, focusClassName, isFocusVisible } = (0, import_useFocusRing.useFocusRing)({
      color,
      disabled
    });
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      "label",
      {
        "data-testid": dataTestId,
        htmlFor: id,
        ref: radioLabelRef,
        className: (0, import_clsx.default)(
          styles.label,
          {
            [styles.disabled]: disabledOrAriaDisabled,
            [styles.checked]: checked
          },
          consumerClassName
        ),
        style: consumerStyle,
        ...behavioralTrackingProps,
        "data-dt-component": COMPONENT_NAME,
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_VisuallyHidden.VisuallyHidden, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "input",
            {
              ref: inputRef,
              checked,
              name: groupName,
              value,
              type: "radio",
              disabled,
              "aria-disabled": ariaDisabled,
              required: groupRequired,
              id,
              onChange: handleChange,
              onInvalid: (e) => {
                if (controlState) {
                  e.preventDefault();
                  return;
                }
                validationOnInvalid(e);
              },
              readOnly,
              ...(0, import_merge_props.mergeProps)(
                restProps,
                focusProps,
                controlState ? {} : {
                  ...ariaProps,
                  onBlur: validationOnBlur
                }
              )
            }
          ) }),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "div",
            {
              className: (0, import_clsx.default)(
                styles.knob({
                  disabled: disabledOrAriaDisabled,
                  checked,
                  isFocusVisible,
                  hasError,
                  readOnly
                }),
                (0, import_field_css.fieldTokens)({
                  disabled: disabledOrAriaDisabled,
                  variant,
                  color
                }),
                focusClassName
              ),
              children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                "svg",
                {
                  className: (0, import_clsx.default)(
                    styles.innerKnob,
                    (0, import_field_css.fieldTokens)({
                      color: hasError ? "critical" : "primary",
                      disabled: disabledOrAriaDisabled,
                      variant: "accent"
                    })
                  ),
                  viewBox: "0 0 100 100",
                  children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("circle", { r: isFocusVisible ? "20" : "25", cx: "50", cy: "50" })
                }
              )
            }
          ),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { as: "span", textStyle: "base", children })
        ]
      }
    );
  }
);
Radio.displayName = COMPONENT_NAME;
