// @dynatrace/strato-components/forms/radio/styles.sty.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var styles_css_exports = {};
__export(styles_css_exports, {
  checked: () => checked,
  disabled: () => disabled,
  group: () => group,
  hasError: () => hasError,
  innerKnob: () => innerKnob,
  knob: () => knob,
  label: () => label,
  readOnly: () => readOnly
});
module.exports = __toCommonJS(styles_css_exports);
var import_AppRoot_css_ts_vanilla = require("../../core/components/app-root/AppRoot.css");
var import_field_css_ts_vanilla = require("../../styles/field.css");
var import_styles_css_ts_vanilla = require("./styles.css");
var import_createRuntimeFn = require("@vanilla-extract/recipes/createRuntimeFn");
var checked = "_h96idu4-3-6-0";
var disabled = "_h96idu3-3-6-0";
var group = "_h96idu0-3-6-0";
var hasError = "_h96idu2-3-6-0";
var innerKnob = "_h96iduk-3-6-0";
var knob = (0, import_createRuntimeFn.createRuntimeFn)({ defaultClassName: "_h96idu6-3-6-0", variantClassNames: { disabled: { true: "_h96idu7-3-6-0" }, checked: { true: "_h96idu8-3-6-0", false: "_h96idu9-3-6-0" }, hasError: { true: "_h96idua-3-6-0", false: "_h96idub-3-6-0" }, isFocusVisible: { true: "_h96iduc-3-6-0", false: "_h96idud-3-6-0" }, readOnly: { true: "_h96idue-3-6-0", false: "_h96iduf-3-6-0" } }, defaultVariants: {}, compoundVariants: [[{ checked: false, hasError: true, isFocusVisible: false, disabled: false, readOnly: false }, "_h96idug-3-6-0"], [{ checked: false, hasError: false, isFocusVisible: false, disabled: false, readOnly: false }, "_h96iduh-3-6-0"], [{ readOnly: true, disabled: false }, "_h96idui-3-6-0"], [{ readOnly: true, hasError: true, disabled: false }, "_h96iduj-3-6-0"]] });
var label = "_h96idu5-3-6-0";
var readOnly = "_h96idu1-3-6-0";
