// @dynatrace/strato-components/charts/core/components/tooltip/section-components/TooltipItem/TooltipItem.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TooltipItem_exports = {};
__export(TooltipItem_exports, {
  TooltipItem: () => TooltipItem
});
module.exports = __toCommonJS(TooltipItem_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_spacings = __toESM(require("@dynatrace/strato-design-tokens/spacings"));
var import_TooltipItem_css = require("./TooltipItem.sty.js");
var import_useId = require("../../../../../../core/hooks/useId.js");
var import_Divider = require("../../../../../../layouts/divider/Divider.js");
var import_ChartActionsMenu = require("../../../series-action/ChartActionsMenu.js");
var import_ActionSelectionContext = require("../../context/ActionSelectionContext.js");
const TooltipItem = (props) => {
  const { children, showDivider, actions, defaultActions, ...rest } = props;
  const itemId = (0, import_useId.useId)();
  const { selectedItemId, actionsMenuRef, updateSelectedItem } = (0, import_ActionSelectionContext.useActionSelection)();
  const isOpen = selectedItemId === itemId;
  const memoizedActions = (0, import_react.useMemo)(() => {
    if (isOpen && actions) {
      return actions();
    }
  }, [isOpen]);
  if (!actions && !defaultActions) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: import_TooltipItem_css.itemContainer, ...rest, children }),
      showDivider && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_Divider.Divider,
        {
          orientation: "horizontal",
          style: { margin: `${import_spacings.default.Size4} 0` }
        }
      )
    ] });
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ChartActionsMenu.ChartActionsMenu,
    {
      isOpen,
      actions: memoizedActions,
      defaultActions,
      onActionClick: () => updateSelectedItem(""),
      ref: actionsMenuRef,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
        "button",
        {
          className: import_TooltipItem_css.container,
          onMouseEnter: () => updateSelectedItem(itemId),
          children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: import_TooltipItem_css.itemContainer, ...rest, children }),
            showDivider && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_Divider.Divider,
              {
                orientation: "horizontal",
                style: { margin: `${import_spacings.default.Size4} 0` }
              }
            )
          ]
        }
      )
    }
  );
};
TooltipItem["displayName"] = "TooltipItem";
