// @dynatrace/strato-components/esm/charts/core/components/overlay-tooltip/overlay.js
const NullPositions = {
  absoluteX: null,
  absoluteY: null,
  relativeX: null,
  relativeY: null,
  percentX: null,
  percentY: null
};
function getAdjustedPositionsFromAbsolute(absoluteX, absoluteY) {
  const bounds = this.getBounds();
  if (!bounds) {
    return NullPositions;
  }
  const { left, top, width, height } = bounds;
  const relativeX = absoluteX - left;
  const relativeY = absoluteY - top;
  const percentX = relativeX * 100 / width;
  const percentY = relativeY * 100 / height;
  return {
    absoluteX,
    absoluteY,
    relativeX,
    relativeY,
    percentX,
    percentY
  };
}
function getAdjustedPositionsFromPercent(percentX, percentY, relativeX, relativeY) {
  const bounds = this.getBounds();
  if (!bounds) {
    return NullPositions;
  }
  const { left, top, width, height } = bounds;
  relativeX ||= percentX * width / 100;
  relativeY ||= percentY * height / 100;
  const absoluteX = Math.round(left + relativeX);
  const absoluteY = Math.round(top + relativeY);
  return {
    absoluteX,
    absoluteY,
    relativeX,
    relativeY,
    percentX,
    percentY
  };
}
function getAdjustedPositionsFromRelative(relativeX, relativeY) {
  const bounds = this.getBounds();
  if (!bounds) {
    return NullPositions;
  }
  const { left, top, width, height } = bounds;
  const percentX = relativeX * 100 / width;
  const percentY = relativeY * 100 / height;
  const absoluteX = Math.round(left + relativeX);
  const absoluteY = Math.round(top + relativeY);
  return {
    absoluteX,
    absoluteY,
    relativeX,
    relativeY,
    percentX,
    percentY
  };
}
function buildOverlay(overlayRef) {
  let bounds = void 0;
  return {
    overlayRef,
    getBounds() {
      bounds ||= overlayRef.current?.getBoundingClientRect();
      return bounds;
    },
    getAdjustedPositionsFromAbsolute,
    getAdjustedPositionsFromPercent,
    getAdjustedPositionsFromRelative,
    clear() {
      bounds = void 0;
    }
  };
}
export {
  buildOverlay
};
