// @dynatrace/strato-components/esm/charts/core/components/overlay-tooltip/OverlayTooltip.js
import { jsx } from "react/jsx-runtime";
import { DynamicTooltipRenderer } from "./components/DynamicTooltipRenderer.js";
import { Footer } from "./components/Footer.js";
import { Action } from "./components/global-actions/Action.js";
import { ActionsPlaceholder } from "./components/global-actions/ActionsPlaceholder.js";
import { Header } from "./components/header/Header.js";
import { Series } from "./components/series/Series.js";
import { SeriesHeader } from "./components/series/SeriesHeader.js";
import { SeriesItem } from "./components/series/SeriesItem.js";
import { TooltipPortal } from "./components/TooltipPortal.js";
const TooltipRoot = (props) => {
  if (props.absoluteX !== void 0 && props.absoluteY !== void 0) {
    return /* @__PURE__ */ jsx(
      TooltipPortal,
      {
        isInteractive: props.isInteractive ?? false,
        absoluteX: props.absoluteX,
        absoluteY: props.absoluteY,
        children: props.children
      }
    );
  }
  return /* @__PURE__ */ jsx(DynamicTooltipRenderer, { ...props });
};
const OverlayTooltip = Object.assign(TooltipRoot, {
  Action,
  ActionsPlaceholder,
  Footer,
  Header,
  Series,
  SeriesHeader,
  SeriesItem
});
const Tooltip = () => null;
Tooltip.displayName = "ChartTooltip";
const ChartTooltip = Object.assign(Tooltip, {
  Action,
  ActionsPlaceholder,
  Footer,
  Header,
  Series,
  SeriesHeader,
  SeriesItem
});
export {
  ChartTooltip,
  OverlayTooltip,
  Tooltip
};
