// @dynatrace/strato-components/esm/charts/core/components/overlay-tooltip/tooltip.sty.js
import "./tooltip.css";
import { createRuntimeFn as _7a468 } from "@vanilla-extract/recipes/createRuntimeFn";
var TOOLTIP_BODY_MAX_HEIGHT = 240;
var TOOLTIP_MAX_WIDTH = 480;
var actionsContainer = "_1j1seex17-3-6-0";
var annotationsGroupsScrollContainer = "_1j1seexc-3-6-0";
var chevron = "_1j1seexl-3-6-0";
var chevronRightAligned = "_1j1seexm-3-6-0";
var divider = "_1j1seex13-3-6-0";
var headerText = "_1j1seexu-3-6-0";
var multiLineName = "_1j1seex12-3-6-0";
var relativeValueStyle = "_1j1seexr-3-6-0";
var seriesActionHeader = "_1j1seex15-3-6-0";
var seriesActionHeaderMaxLinesVar = "var(--_1j1seex2-3-6-0)";
var seriesActionHeaderMultiLine = "_1j1seex16-3-6-0";
var seriesDescriptionText = "_1j1seexz-3-6-0";
var seriesDescriptionTextMultiLine = "_1j1seex10-3-6-0";
var seriesDescriptionTextMultiLineWrap = "_1j1seex11-3-6-0";
var seriesGridContainer = "_1j1seexb-3-6-0";
var seriesHeader = "_1j1seex7-3-6-0";
var seriesHeaderName = "_1j1seex8-3-6-0";
var seriesHeaderValue = "_1j1seex9-3-6-0";
var seriesItem = _7a468({ defaultClassName: "_1j1seexf-3-6-0", variantClassNames: { symbolPosition: { left: "_1j1seexg-3-6-0", right: "_1j1seexh-3-6-0" }, hasChevron: { true: "_1j1seexi-3-6-0" } }, defaultVariants: { symbolPosition: "left", hasChevron: false }, compoundVariants: [] });
var seriesItemActive = "_1j1seexn-3-6-0";
var seriesItemHighlighted = "_1j1seex18-3-6-0";
var seriesItemMaxWidthVar = "var(--_1j1seex1-3-6-0)";
var seriesItemText = "_1j1seexo-3-6-0";
var seriesItemWithChevron = "_1j1seexk-3-6-0";
var seriesItemWrapper = "_1j1seexj-3-6-0";
var seriesNameCell = "_1j1seexx-3-6-0";
var seriesNameMaxLinesVar = "var(--_1j1seex0-3-6-0)";
var seriesNameText = "_1j1seexv-3-6-0";
var seriesNameTextMultiLine = "_1j1seexp-3-6-0";
var seriesNameTextMultiLineWrap = "_1j1seexq-3-6-0";
var seriesOverflowContainer = "_1j1seexe-3-6-0";
var seriesOverflowContainerBase = "_1j1seexd-3-6-0";
var seriesValueCell = "_1j1seexy-3-6-0";
var seriesValueText = "_1j1seexw-3-6-0";
var seriesWrapper = "_1j1seexa-3-6-0";
var spaceSavedForSymbolVar = "var(--_1j1seex3-3-6-0)";
var symbolRight = "_1j1seex14-3-6-0";
var symbolWrapper = _7a468({ defaultClassName: "_1j1seexs-3-6-0", variantClassNames: { isRight: { true: "_1j1seext-3-6-0" } }, defaultVariants: {}, compoundVariants: [] });
var wrapper = _7a468({ defaultClassName: "_1j1seex4-3-6-0", variantClassNames: { isInteractive: { true: "_1j1seex5-3-6-0", false: "_1j1seex6-3-6-0" } }, defaultVariants: {}, compoundVariants: [] });
export {
  TOOLTIP_BODY_MAX_HEIGHT,
  TOOLTIP_MAX_WIDTH,
  actionsContainer,
  annotationsGroupsScrollContainer,
  chevron,
  chevronRightAligned,
  divider,
  headerText,
  multiLineName,
  relativeValueStyle,
  seriesActionHeader,
  seriesActionHeaderMaxLinesVar,
  seriesActionHeaderMultiLine,
  seriesDescriptionText,
  seriesDescriptionTextMultiLine,
  seriesDescriptionTextMultiLineWrap,
  seriesGridContainer,
  seriesHeader,
  seriesHeaderName,
  seriesHeaderValue,
  seriesItem,
  seriesItemActive,
  seriesItemHighlighted,
  seriesItemMaxWidthVar,
  seriesItemText,
  seriesItemWithChevron,
  seriesItemWrapper,
  seriesNameCell,
  seriesNameMaxLinesVar,
  seriesNameText,
  seriesNameTextMultiLine,
  seriesNameTextMultiLineWrap,
  seriesOverflowContainer,
  seriesOverflowContainerBase,
  seriesValueCell,
  seriesValueText,
  seriesWrapper,
  spaceSavedForSymbolVar,
  symbolRight,
  symbolWrapper,
  wrapper
};
