// @dynatrace/strato-components/esm/charts/core/components/overlay-tooltip/useTooltipActions.js
import { useCallback } from "react";
import { useTooltip } from "./contexts/TooltipContext.js";
import { useTooltipReducer } from "./useTooltipReducer.js";
import { emptyData } from "./utils.js";
const useTooltipActions = () => {
  const dispatch = useTooltipReducer();
  const tooltipState = useTooltip();
  const hover = useCallback(
    (event, data) => {
      dispatch({
        type: "SET_POSITION",
        payload: {
          absoluteX: Math.round(event.clientX),
          absoluteY: Math.round(event.clientY),
          data
        }
      })();
    },
    [dispatch]
  );
  const pin = useCallback(
    (event, data) => {
      const { absoluteX, absoluteY } = tooltipState.getState();
      const coords = absoluteX !== null && absoluteY !== null ? { clientX: absoluteX, clientY: absoluteY } : {
        clientX: Math.round(event.clientX),
        clientY: Math.round(event.clientY)
      };
      dispatch({
        type: "SET_PINNED",
        payload: {
          pinned: true,
          absoluteX: coords.clientX,
          absoluteY: coords.clientY,
          data
        }
      })();
    },
    [dispatch, tooltipState]
  );
  const unpin = useCallback(() => {
    dispatch({
      type: "SET_PINNED",
      payload: { pinned: false, data: emptyData() }
    })();
  }, [dispatch]);
  const togglePin = useCallback(
    (event, data) => {
      if (tooltipState.getState().pinned) {
        unpin();
      } else {
        pin(event, data);
      }
    },
    [pin, unpin, tooltipState]
  );
  const reset = useCallback(() => {
    dispatch({ type: "RESET_TOOLTIP" })();
  }, [dispatch]);
  const outOfBounds = useCallback(() => {
    dispatch({ type: "SET_OUT_BOUNDS" })();
  }, [dispatch]);
  return { hover, pin, togglePin, unpin, reset, outOfBounds };
};
export {
  useTooltipActions
};
