// @dynatrace/strato-components/charts/pie/components/InnerWrapper/components/InnerDonut/InnerDonut.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var InnerDonut_exports = {};
__export(InnerDonut_exports, {
  InnerDonut: () => InnerDonut
});
module.exports = __toCommonJS(InnerDonut_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_InnerDonut_css = require("./InnerDonut.sty.js");
var import_Flex = require("../../../../../../layouts/flex/Flex.js");
var import_InnerElementContext = require("../../../../context/InnerElementContext.js");
var import_selectors = require("../../../../hooks/store/selectors.js");
var import_build_cumulative_value = require("../../../../utils/build-cumulative-value.js");
const InnerDonut = ({ donutInner, visibleIds }) => {
  const { innerElement } = (0, import_InnerElementContext.useChartInnerElement)();
  const visibleSlices = (0, import_selectors.useVisibleSlicesState)();
  const { absoluteValue, relativeValue } = (0, import_react.useMemo)(
    () => (0, import_build_cumulative_value.buildCumulativeValues)(visibleSlices, visibleIds),
    [visibleIds, visibleSlices]
  );
  if (!innerElement) {
    return null;
  }
  const containerStyle = {
    width: `${donutInner}px`,
    height: `${donutInner}px`
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Flex.Flex,
    {
      justifyContent: "center",
      alignItems: "center",
      className: import_InnerDonut_css.container,
      style: containerStyle,
      children: (0, import_lodash_es.isFunction)(innerElement) ? innerElement({
        absoluteValue,
        relativeValue: (0, import_lodash_es.round)(relativeValue, 2),
        dimensions: { width: donutInner, height: donutInner }
      }) : innerElement
    }
  );
};
