// @dynatrace/strato-components/charts/pie/components/PieChartLegend/PieChartLegend.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var PieChartLegend_exports = {};
__export(PieChartLegend_exports, {
  PieChartLegend: () => PieChartLegend
});
module.exports = __toCommonJS(PieChartLegend_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_useChartLayoutLoading = require("../../../core/components/chart-layout/hooks/useChartLayoutLoading.js");
var import_useChartLayoutOrientation = require("../../../core/components/chart-layout/hooks/useChartLayoutOrientation.js");
var import_useChartLayoutSize = require("../../../core/components/chart-layout/hooks/useChartLayoutSize.js");
var import_orientation = require("../../../core/components/chart-layout/utils/orientation.js");
var import_ChartLegend = require("../../../core/components/legend/categorical/ChartLegend.js");
var import_useLegendSeriesActions = require("../../../core/components/legend/categorical/hooks/useLegendSeriesActions.js");
var import_constants = require("../../../core/constants.js");
var import_FormatterContext = require("../../context/FormatterContext.js");
var import_PieChartSlicesData = require("../../context/PieChartSlicesData.js");
var import_usePieHandlers = require("../../hooks/usePieHandlers.js");
const PieChartLegend = (0, import_react.memo)(() => {
  const { slices } = (0, import_PieChartSlicesData.usePieChartSlicesData)();
  const { height, width, wrapperHeight } = (0, import_useChartLayoutSize.useChartLayoutSize)();
  const { orientation: layoutOrientation } = (0, import_useChartLayoutOrientation.useChartLayoutOrientation)();
  const hideLegendForAspectRatio = layoutOrientation === import_constants.ORIENTATION.Horizontal ? width < height : wrapperHeight < height;
  const { onLegendItemHover, onLegendItemClicked } = (0, import_usePieHandlers.usePieHandlers)(slices);
  const legendOrientation = (0, import_orientation.getLegendOrientation)(layoutOrientation);
  const truncationMode = (0, import_FormatterContext.useTruncationMode)();
  const legendActions = (0, import_useLegendSeriesActions.useLegendSeriesActions)(
    slices,
    (s) => s.id
  );
  const isLoading = (0, import_useChartLayoutLoading.useChartLayoutLoading)();
  const legendItems = slices.map((slice) => ({
    id: slice.id,
    color: slice.color,
    name: slice.category
  }));
  return !hideLegendForAspectRatio && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ChartLegend.ChartLegend,
    {
      items: legendItems,
      orientation: legendOrientation,
      onHover: onLegendItemHover,
      onClick: onLegendItemClicked,
      onSeriesAction: legendActions,
      truncationMode,
      interactionsDisabled: isLoading
    }
  );
});
PieChartLegend["displayName"] = "PieChartLegend";
