// @dynatrace/strato-design-tokens/borders/index.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var borders_exports = {};
__export(borders_exports, {
  default: () => borders_default
});
module.exports = __toCommonJS(borders_exports);
const Style = {
  /**
   * Use as the default border style.
   * Css custom property: --dt-borders-style-default. Default value: solid.
   */
  Default: "var(--dt-borders-style-default, solid)",
  /**
   * Use to indicate interactive text.
   * Css custom property: --dt-borders-style-dashed. Default value: dashed.
   */
  Dashed: "var(--dt-borders-style-dashed, dashed)"
};
const Width = {
  /**
   * Use as the default border width.
   * Css custom property: --dt-borders-width-default. Default value: 1px.
   */
  Default: "var(--dt-borders-width-default, 1px)",
  /**
   * Use to emphasize borders. This can be useful to clearly separate content.
   * Css custom property: --dt-borders-width-emphasized. Default value: 2px.
   */
  Emphasized: "var(--dt-borders-width-emphasized, 2px)"
};
const Radius = {
  Surface: {
    /**
     * Use as the subdued border radius for surfaces.
     * Css custom property: --dt-borders-radius-surface-subdued. Default value: 9px.
     */
    Subdued: "var(--dt-borders-radius-surface-subdued, 9px)",
    /**
     * Use as the default border radius for surfaces.
     * Css custom property: --dt-borders-radius-surface-default. Default value: 12px.
     */
    Default: "var(--dt-borders-radius-surface-default, 12px)"
  },
  Container: {
    /**
     * Use as the subdued border radius for containers.
     * Css custom property: --dt-borders-radius-container-subdued. Default value: 6px.
     */
    Subdued: "var(--dt-borders-radius-container-subdued, 6px)",
    /**
     * Use as the default border radius for containers.
     * Css custom property: --dt-borders-radius-container-default. Default value: 9px.
     */
    Default: "var(--dt-borders-radius-container-default, 9px)"
  },
  Field: {
    /**
     * Use as the subdued border radius for fields.
     * Css custom property: --dt-borders-radius-field-subdued. Default value: 3px.
     */
    Subdued: "var(--dt-borders-radius-field-subdued, 3px)",
    /**
     * Use as the default border radius for fields.
     * Css custom property: --dt-borders-radius-field-default. Default value: 6px.
     */
    Default: "var(--dt-borders-radius-field-default, 6px)",
    /**
     * Use for rounded or pill-like shapes.
     * Css custom property: --dt-borders-radius-field-emphasized. Default value: 9999px.
     */
    Emphasized: "var(--dt-borders-radius-field-emphasized, 9999px)"
  }
};
var borders_default = {
  Style,
  Width,
  Radius
};
