// @dynatrace/strato-components/tables/utils/log-content-highlighting.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var log_content_highlighting_exports = {};
__export(log_content_highlighting_exports, {
  LogContentHighlight: () => LogContentHighlight
});
module.exports = __toCommonJS(log_content_highlighting_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_error_boundary = require("react-error-boundary");
var import_HighlightExtended = require("../../content/code-snippet/HighlightExtended.js");
var import_sdk_logger = require("../../core/sdk/sdk-logger.js");
var import_textStyle_css = require("../../styles/textStyle.sty.js");
const ELLIPSIS_CHARACTER = "\u2026";
const MAX_TOKENIZATION_LENGTH = 1e3;
const DEFAULT_TRUNCATION_LENGTH = MAX_TOKENIZATION_LENGTH;
function useFormatLogContent(logContent, options) {
  const stratoLogger = (0, import_sdk_logger.getStratoLogger)("DataTable");
  return (0, import_react.useMemo)(() => {
    if (logContent !== null && logContent !== void 0) {
      const lineWrap = options?.lineWrap ?? true;
      let logContentValue = logContent;
      if (!lineWrap) {
        logContentValue = logContent.replace(/[\n\r]/g, " ");
      }
      const truncationLimit = getTruncationLimit(options?.truncationLimit);
      const highlightIndex = Math.min(
        MAX_TOKENIZATION_LENGTH,
        truncationLimit,
        // in case truncation limit is lower than 1000
        logContentValue.length
        // in case log line is shorter than 1000 chars
      );
      let logContentSubstrings = {
        // max. first 1000 characters are tokenized and highlighted
        highlight: logContentValue.substring(0, highlightIndex),
        // remaining characters that are neither tokenized nor highlighted to enhance performance
        noHighlight: logContentValue.substring(
          highlightIndex,
          Math.min(truncationLimit, logContentValue.length)
        )
      };
      if (truncationLimit < logContentValue.length) {
        logContentSubstrings = ellipsize(logContentSubstrings);
      }
      const noHighlightLines = logContentSubstrings.noHighlight.length > 0 ? logContentSubstrings.noHighlight.split(/\r?\n/) : [];
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_react_error_boundary.ErrorBoundary,
        {
          fallback: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { children: logContentValue.substring(0, truncationLimit) }),
          onError: (error, info) => {
            stratoLogger.error(error.message, error, {
              logContentValue,
              info
            });
          },
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_HighlightExtended.HighlightExtended,
            {
              language: "log",
              code: logContentSubstrings.highlight,
              children: ({ tokens, getLineProps, getTokenProps, className, style }) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, { children: tokens.map((line, i) => {
                const { className: lineClassName, ...remainingLineProps } = getLineProps({ line });
                return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
                  "div",
                  {
                    ...remainingLineProps,
                    className: (0, import_clsx.default)(
                      className,
                      (0, import_textStyle_css.textStyle)({ fontStyle: "code" })
                    ),
                    style: { ...style, width: "100%" },
                    children: [
                      line.map((token, j) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { ...getTokenProps({ token }) }, j)),
                      /* render the remaining content that is not highlighted
                         in the same `token-line` after the last token  */
                      i === tokens.length - 1 ? noHighlightLines.map((line2, j) => /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_react.Fragment, { children: [
                        /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { children: line2 }),
                        j < noHighlightLines.length - 1 ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)("br", {}) : null
                      ] }, `${i}-${j}`)) : null
                    ]
                  },
                  i
                );
              }) })
            },
            logContentSubstrings.highlight
          )
        }
      );
    }
    return null;
  }, [logContent, options?.lineWrap, options?.truncationLimit, stratoLogger]);
}
function getTruncationLimit(truncationLimit) {
  if (truncationLimit === void 0) {
    return DEFAULT_TRUNCATION_LENGTH;
  }
  if (truncationLimit === false) {
    return Infinity;
  }
  return truncationLimit;
}
function ellipsize(content) {
  let { highlight, noHighlight } = content;
  const trimmedNoHighlight = noHighlight.trimEnd();
  if (trimmedNoHighlight.length > 0) {
    noHighlight = trimmedNoHighlight + ELLIPSIS_CHARACTER;
  } else {
    const trimmedHighlight = content.highlight.trimEnd();
    if (trimmedHighlight.length > 0) {
      highlight = trimmedHighlight + ELLIPSIS_CHARACTER;
    }
  }
  return { highlight, noHighlight };
}
function LogContentHighlight(props) {
  const { logContent, lineWrap, truncationLimit } = props;
  return useFormatLogContent(logContent, { lineWrap, truncationLimit });
}
