// @dynatrace/strato-components/esm/content/keyboard-shortcut-tooltip/KeyboardShortcutTooltip.js
import { jsx, jsxs } from "react/jsx-runtime";
import { forwardRef } from "react";
import Typography from "@dynatrace/strato-design-tokens/typography";
import { useAriaLabelingProps } from "../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useSafeForwardProps } from "../../core/hooks/useSafeForwardProps.js";
import { Flex } from "../../layouts/flex/Flex.js";
import { Tooltip } from "../../overlays/tooltip/Tooltip.js";
import { Text } from "../../typography/text/Text.js";
import { KeyboardShortcut } from "../keyboard-shortcut/KeyboardShortcut.js";
const COMPONENT_NAME = "KeyboardShortcutTooltip";
const KeyboardShortcutTooltip = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const {
    children,
    keys,
    shortcutDisabled,
    text,
    placement = "top",
    open,
    defaultOpen,
    onOpenChange,
    delay = "default",
    disabled,
    style: consumerStyle,
    className: consumerClassName,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
  const ariaLabelingProps = useAriaLabelingProps(restProps);
  const forwardProps = useSafeForwardProps(restProps, ariaLabelingProps);
  return /* @__PURE__ */ jsx(
    Tooltip,
    {
      ref: forwardedRef,
      placement,
      open,
      defaultOpen,
      onOpenChange,
      delay,
      disabled,
      text: /* @__PURE__ */ jsxs(Flex, { flexDirection: "column", alignItems: "center", gap: 4, children: [
        text && /* @__PURE__ */ jsx(
          Text,
          {
            textStyle: "small-emphasized",
            style: { maxWidth: "100%", overflowWrap: "anywhere" },
            children: text
          }
        ),
        /* @__PURE__ */ jsx(
          KeyboardShortcut,
          {
            keys,
            variant: "minimal",
            disabled: shortcutDisabled,
            style: {
              lineHeight: Typography.Text.Small.Default.LineHeight
            }
          }
        )
      ] }),
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...ariaLabelingProps,
      ...behavioralTrackingProps,
      ...forwardProps,
      "data-dt-component": COMPONENT_NAME,
      children
    }
  );
});
KeyboardShortcutTooltip.displayName = COMPONENT_NAME;
export {
  KeyboardShortcutTooltip
};
