// @dynatrace/strato-components/esm/charts/histogram/components/axes/HistogramValueBottomAxisArea/HistogramValueBottomAxisArea.js
import { jsx, jsxs } from "react/jsx-runtime";
import { useGesture } from "@use-gesture/react";
import clsx from "clsx";
import { isEmpty } from "lodash-es";
import { useRef } from "react";
import { HistogramPanningControls } from "../HistogramPanningControls.js";
import { axisContainerStyle } from "./HistogramValueBottomAxisArea.sty.js";
import { ValueBottomAxis } from "../../../../core/components/axes/components/ValueBottomAxis.js";
import { ChartGrid } from "../../../../core/components/chart-grid/ChartGridRoot.js";
import { KEYBOARD_HANDLERS } from "../../../../core/components/zoom/constants.js";
import { cursorStyle } from "../../../../core/components/zoom/ZoomCursors.sty.js";
import { AREAS } from "../../../constants/histogram-grid-area-rules.js";
import { useZoomState } from "../../../hooks/store/selectors.js";
import { useHistogramInteractions } from "../../../hooks/useHistogramInteractions.js";
import { useHistogramTickAxis } from "../../../hooks/useHistogramTickAxis.js";
import { useHistogramXScale } from "../../../hooks/useHistogramXScale.js";
import { useXAxisConfig } from "../../../hooks/useXAxisConfig.js";
import { useApplyPanDelta } from "../../overlays/hooks/useApplyPanDelta.js";
const HistogramValueBottomAxisArea = () => {
  const axisRef = useRef(null);
  const { label } = useXAxisConfig();
  const ticks = useHistogramTickAxis();
  const { pan } = useHistogramInteractions();
  const { zoomOutEnabled } = useZoomState();
  const { xScale } = useHistogramXScale();
  const isPanEnabled = pan?.enabled && zoomOutEnabled;
  const applyPan = useApplyPanDelta();
  useGesture(
    {
      onDrag: (e) => !KEYBOARD_HANDLERS.includes(e.type) && applyPan(-e.delta[0])
    },
    {
      target: axisRef,
      eventOptions: { passive: false },
      drag: { filterTaps: true, pointer: { buttons: [1, 4] } },
      enabled: isPanEnabled
    }
  );
  const bottomTicks = ticks?.bottom || { ticks: [], formattedTicks: [] };
  if (isEmpty(bottomTicks.ticks) || !xScale) {
    return null;
  }
  return /* @__PURE__ */ jsx(ChartGrid.GridArea, { gridArea: AREAS.bottomAxis, children: ({ availableWidth, availableHeight }) => {
    return /* @__PURE__ */ jsxs(
      "div",
      {
        className: clsx(cursorStyle({ isPanEnabled }), axisContainerStyle),
        ref: axisRef,
        "data-testid": "histogram-bottom-axis",
        children: [
          isPanEnabled && /* @__PURE__ */ jsx(HistogramPanningControls, { onClickPan: (d) => applyPan(-d) }),
          /* @__PURE__ */ jsx(
            ValueBottomAxis,
            {
              ticks: bottomTicks.ticks,
              formattedTicks: bottomTicks.formattedTicks,
              containerHeight: availableHeight,
              containerWidth: availableWidth,
              scale: xScale,
              label
            }
          )
        ]
      }
    );
  } });
};
HistogramValueBottomAxisArea["displayName"] = "HistogramValueBottomAxisArea";
export {
  HistogramValueBottomAxisArea
};
