// @dynatrace/strato-components/esm/typography/external-link/ExternalLink.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import { forwardRef } from "react";
import { useIntl } from "react-intl";
import { ExternalLinkIcon } from "@dynatrace/strato-icons";
import * as styles from "./ExternalLink.sty.js";
import { useAriaLabelingProps } from "../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useSafeForwardProps } from "../../core/hooks/useSafeForwardProps.js";
const COMPONENT_NAME = "ExternalLink";
const ExternalLink = /* @__PURE__ */ forwardRef((props, ref) => {
  const {
    children,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
  const ariaLabelingProps = useAriaLabelingProps(restProps);
  const forwardProps = useSafeForwardProps(restProps, ariaLabelingProps);
  const intl = useIntl();
  return /* @__PURE__ */ jsxs(
    "a",
    {
      ref,
      target: "_blank",
      rel: "noopener noreferrer",
      className: clsx(styles.externalLink, consumerClassName),
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...ariaLabelingProps,
      ...behavioralTrackingProps,
      ...forwardProps,
      "data-dt-component": COMPONENT_NAME,
      children: [
        children,
        /* @__PURE__ */ jsxs("span", { style: { userSelect: "none" }, children: [
          "\uFEFF",
          /* @__PURE__ */ jsx(
            ExternalLinkIcon,
            {
              "aria-hidden": "false",
              "aria-label": intl.formatMessage({
                defaultMessage: "Opening the link in a new window",
                id: "JzYy8MUVfPcyf4l5",
                description: "External link icon that describes the external link icon"
              })
            }
          )
        ] })
      ]
    }
  );
});
ExternalLink.displayName = COMPONENT_NAME;
export {
  ExternalLink
};
