// @dynatrace/strato-components/forms/date-time-range-picker/hooks/useDateTimeInput.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useDateTimeInput_exports = {};
__export(useDateTimeInput_exports, {
  useDateTimeInput: () => useDateTimeInput
});
module.exports = __toCommonJS(useDateTimeInput_exports);
var import_tz = require("@date-fns/tz");
var import_react = require("react");
var import_units = require("@dynatrace-sdk/units");
var import_user_preferences = require("@dynatrace-sdk/user-preferences");
var import_useDateTimeParser = require("./useDateTimeParser.js");
var import_useExpressionSpin = require("./useExpressionSpin.js");
var import_useAriaLabelingProps = require("../../../core/hooks/useAriaLabelingProps.js");
var import_useControllableState = require("../../../core/hooks/useControllableState.js");
var import_helpers = require("../../../core/time/helpers.js");
var import_useTimeframeStore = require("../../../core/time/timeframeStore/useTimeframeStore.js");
var import_useDateTimePicker = require("../../date-time-picker/hooks/useDateTimePicker.js");
var import_timeParser = require("../../forms-core/time/timeParser.js");
var import_useSelection = require("../../spin-buttons/hooks/useSelection.js");
var import_helpers2 = require("../helpers.js");
var import_selection_helpers = require("../selection-helpers.js");
function isNumberSpinButton(handler) {
  return handler?.type === "number";
}
const DATE_PART_TYPES = /* @__PURE__ */ new Set([
  "day",
  "month",
  "year"
]);
const TIME_PART_TYPES = /* @__PURE__ */ new Set([
  "hour",
  "minute",
  "second",
  "fractionalSecond",
  "dayPeriod"
]);
const insertTextAtCursor = (text) => {
  const selection = window.getSelection();
  if (selection?.rangeCount) {
    const range = selection.getRangeAt(0);
    range.deleteContents();
    const textNode = document.createTextNode(text);
    range.insertNode(textNode);
    range.setStartAfter(textNode);
    range.collapse(true);
    selection.removeAllRanges();
    selection.addRange(range);
  }
};
const hasAnyEmptySpinButton = (handlersRef) => Array.from(handlersRef.current?.entries() ?? []).some(
  ([type, handler]) => handler && type !== "literal" && handler.element.current?.textContent === ""
);
const pasteIntoExpressionInput = (element, text) => {
  if (element.current) {
    const selection = window.getSelection();
    const range = selection?.rangeCount ? selection.getRangeAt(0) : null;
    if (range && element.current.contains(range.startContainer) && element.current.contains(range.endContainer)) {
      insertTextAtCursor(text);
    } else {
      element.current.textContent = text;
      moveCursorToEnd(element.current);
    }
    element.current.dispatchEvent(new InputEvent("input", { bubbles: true }));
  }
};
const moveCursorToEnd = (element) => {
  if (element) {
    const range = document.createRange();
    const selection = window.getSelection();
    range.selectNodeContents(element);
    range.collapse(false);
    selection?.removeAllRanges();
    selection?.addRange(range);
  }
};
const useDateTimeInput = (props, expressionInputRef) => {
  const {
    disabled,
    readOnly,
    type,
    onChange,
    onLeadingEdge,
    onTrailingEdge,
    dateOnlyPasteAnchor = "startOfDay",
    onPasteComplete,
    partialTimeAnchorDate
  } = props;
  const precision = (0, import_useTimeframeStore.useDateTimePickerStore)(
    (state) => state.precision !== "day" ? state.precision : "milliseconds"
  );
  const ariaProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(props);
  const { regexDateTimeStart, maxValue, types } = (0, import_useDateTimeParser.useDateTimeParser)({
    type
  });
  const [value, setValue] = (0, import_useControllableState.useControllableState)(props);
  const isControlled = props.value !== void 0;
  const [valueType, setValueType] = (0, import_react.useState)(value?.type ?? "expression");
  const firstRun = (0, import_react.useRef)(true);
  const prevValueRef = (0, import_react.useRef)(void 0);
  const [expressionEmpty, setExpressionEmpty] = (0, import_react.useState)(
    !expressionInputRef.current?.textContent
  );
  const [inputValue, setInputValue] = (0, import_react.useState)("");
  const pasteUndoRef = (0, import_react.useRef)(null);
  (0, import_react.useEffect)(() => {
    return () => {
      pasteUndoRef.current = null;
    };
  }, []);
  const lastValidAbsoluteDateRef = (0, import_react.useRef)(
    value?.type === "iso8601" && value.absoluteDate ? value.absoluteDate : null
  );
  const isSwitchingToExpressionRef = (0, import_react.useRef)(false);
  if (valueType === "iso8601" && value?.type === "iso8601" && value.absoluteDate) {
    lastValidAbsoluteDateRef.current = value.absoluteDate;
  }
  const spinButtonValue = valueType === "iso8601" ? lastValidAbsoluteDateRef.current : null;
  const pendingFocusPartsRef = (0, import_react.useRef)(null);
  const {
    spinButtons,
    handlers: spinButtonHandlers,
    empty: spinButtonsEmpty,
    applyDateTime,
    firstSpinButtonHandler
  } = (0, import_useDateTimePicker.useDateTimePicker)({
    type,
    precision,
    disabled,
    readOnly,
    valueType,
    value: spinButtonValue,
    onChange: (newIsoDateTime) => {
      const newValue = (0, import_timeParser.getTimeValue)(newIsoDateTime ?? void 0, precision);
      if (newValue === void 0) {
        return;
      }
      if (isSwitchingToExpressionRef.current) {
        return;
      }
      if (expressionInputRef.current) {
        expressionInputRef.current.textContent = "";
      }
      if (valueType !== "iso8601") {
        setValueType("iso8601");
      }
      if (newIsoDateTime === "" || newValue?.absoluteDate === "") {
        onChange?.(newValue);
        return;
      }
      setValue(
        (prevValue) => (0, import_helpers.isEqualTimeValue)(prevValue, newValue) ? prevValue : newValue
      );
    },
    ...ariaProps
  });
  const empty = (0, import_react.useMemo)(
    () => valueType === "expression" ? expressionEmpty : spinButtonsEmpty,
    [valueType, expressionEmpty, spinButtonsEmpty]
  );
  const resetToExpressionMode = (0, import_react.useCallback)(
    (expressionEmpty2 = true) => {
      setValueType("expression");
      setValue(null);
      setInputValue("");
      setExpressionEmpty(expressionEmpty2);
      lastValidAbsoluteDateRef.current = null;
    },
    [setValue]
  );
  const prevAbsoluteDateRef = (0, import_react.useRef)(null);
  const formatExpressionValue = (0, import_react.useCallback)(
    (value2) => {
      if (!expressionInputRef.current) {
        return;
      }
      const formattedValue = value2 ? (0, import_timeParser.formatTimeValue)(value2) : "";
      const currentValue = (0, import_timeParser.formatTimeValue)(
        expressionInputRef.current?.textContent ?? ""
      );
      if (currentValue !== formattedValue) {
        expressionInputRef.current.textContent = formattedValue;
      }
    },
    // no use in adding the expressionInputRef to the dep. array
    // eslint-disable-next-line react-hooks/exhaustive-deps
    []
  );
  const spinButtonHandlersRef = (0, import_react.useRef)(spinButtonHandlers);
  spinButtonHandlersRef.current = spinButtonHandlers;
  (0, import_react.useEffect)(() => {
    const prevValue = prevValueRef.current;
    prevValueRef.current = value;
    if (value?.type === "iso8601" && !spinButtonsEmpty) {
      setInputValue(value.value);
    }
    if (!isControlled || typeof value === "string") {
      return;
    }
    const prevAbsoluteDate = prevAbsoluteDateRef.current;
    prevAbsoluteDateRef.current = value?.absoluteDate ?? null;
    if (!value) {
      setValueType("expression");
      formatExpressionValue(null);
      const input = expressionInputRef.current?.textContent;
      setExpressionEmpty(input === "");
      spinButtonHandlers.forEach((sb) => sb?.clearValue(true));
      lastValidAbsoluteDateRef.current = null;
      return;
    }
    if (value.absoluteDate === "") {
      return;
    }
    setValueType(value.type);
    setInputValue(value.value);
    if (value.type === "expression") {
      if (!(0, import_helpers.isEqualTimeValue)(prevValue ?? null, value)) {
        formatExpressionValue(value);
      }
      setExpressionEmpty(value.value === "");
    } else if (value.type === "iso8601" && hasAnyEmptySpinButton(spinButtonHandlersRef) && prevAbsoluteDate !== value.absoluteDate) {
      applyDateTime(value.absoluteDate);
    }
  }, [value]);
  const focusFirstUnfilledSpinButton = (0, import_react.useCallback)(
    (filledParts) => {
      const handlers = spinButtonHandlersRef.current;
      for (const [type2, handler] of handlers) {
        if (handler && type2 !== "literal" && !filledParts.has(
          type2
        ) && handler.element.current?.textContent === "") {
          handler.element.current?.focus();
          return;
        }
      }
      let lastHandler = null;
      for (const [, handler] of handlers) {
        if (handler) {
          lastHandler = handler;
        }
      }
      lastHandler?.element?.current?.focus();
    },
    []
  );
  (0, import_react.useEffect)(() => {
    const filledParts = pendingFocusPartsRef.current;
    if (!filledParts || valueType !== "iso8601") {
      return;
    }
    pendingFocusPartsRef.current = null;
    requestAnimationFrame(() => {
      focusFirstUnfilledSpinButton(filledParts);
    });
  }, [valueType, focusFirstUnfilledSpinButton]);
  const { selectAll: expressionInputSelectAll } = (0, import_useSelection.useSelection)(expressionInputRef);
  const expressionInputHandler = (0, import_react.useCallback)(() => {
    if (!expressionInputRef.current) {
      return;
    }
    const input = expressionInputRef.current.textContent ?? "";
    if (expressionInputRef.current.innerHTML === "<br>") {
      expressionInputRef.current.innerHTML = "";
    }
    if (input === "") {
      setValueType("expression");
      setValue(null);
      setExpressionEmpty(true);
      return;
    }
    setExpressionEmpty(false);
    const [, carryOver] = regexDateTimeStart.exec(input) ?? [];
    if (Number.parseInt(carryOver) <= maxValue) {
      setValueType("iso8601");
      applyDateTime(null);
      lastValidAbsoluteDateRef.current = null;
      const sortedHandlers = types.filter((type2) => type2 !== "literal").map((type2) => spinButtonHandlers.get(type2)).filter(Boolean);
      const [carryOverSB, focusSB] = sortedHandlers;
      isNumberSpinButton(carryOverSB) && carryOverSB.applyText(carryOver, true);
      expressionInputRef.current.textContent = "";
      requestAnimationFrame(() => {
        focusSB?.element.current?.focus();
      });
      return;
    }
    if (type !== "date") {
      const colonMatch = /^(\d{1,2}):$/.exec(input);
      if (colonMatch) {
        const hour = Number.parseInt(colonMatch[1]);
        if (hour <= 23) {
          const base = partialTimeAnchorDate ? new import_tz.TZDateMini(partialTimeAnchorDate, (0, import_user_preferences.getTimezone)()) : import_tz.TZDateMini.tz((0, import_user_preferences.getTimezone)());
          const iso = import_tz.TZDateMini.tz(
            (0, import_user_preferences.getTimezone)(),
            base.getFullYear(),
            base.getMonth(),
            base.getDate(),
            hour,
            0,
            0,
            0
          ).toISOString();
          const partsToUpdate = /* @__PURE__ */ new Set(["day", "month", "year", "hour"]);
          setValueType("iso8601");
          expressionInputRef.current.textContent = "";
          applyDateTime(iso, partsToUpdate);
          requestAnimationFrame(() => {
            focusFirstUnfilledSpinButton(partsToUpdate);
          });
          return;
        }
      }
    }
    const typedPartialDate = (0, import_helpers2.tryParsePartialDate)(input);
    if (typedPartialDate) {
      const dateResult = (0, import_helpers2.partialDateTimeToISO)(typedPartialDate, null);
      if (dateResult) {
        setValueType("iso8601");
        expressionInputRef.current.textContent = "";
        applyDateTime(dateResult.iso, dateResult.partsToUpdate);
        requestAnimationFrame(
          () => focusFirstUnfilledSpinButton(dateResult.partsToUpdate ?? /* @__PURE__ */ new Set())
        );
        return;
      }
    }
    const newValue = (0, import_timeParser.getTimeValue)(input, precision) ?? null;
    if (newValue?.absoluteDate === "") {
      onChange?.(newValue);
    }
    setValue(
      (prevValue) => (0, import_helpers.isEqualTimeValue)(prevValue, newValue) ? prevValue : newValue
    );
    setInputValue(input);
  }, [
    expressionInputRef,
    regexDateTimeStart,
    maxValue,
    precision,
    type,
    setValue,
    applyDateTime,
    focusFirstUnfilledSpinButton,
    types,
    spinButtonHandlers,
    onChange
  ]);
  const expressionOnFocusHandler = (0, import_react.useCallback)(
    (event) => {
      const selection = window.getSelection();
      const selectionSpansMultipleNodes = selection !== null && selection.anchorNode !== selection.focusNode && selection.toString().length > 0;
      const isRestoringSelection = event.target.closest(".strato-inputgroup")?.hasAttribute("data-restoring-selection") ?? false;
      const expressionInputEl = event.target;
      const hasPartialSelectionWithin = selection !== null && !selection.isCollapsed && expressionInputEl.contains(selection.anchorNode) && expressionInputEl.contains(selection.focusNode);
      if (selectionSpansMultipleNodes || isRestoringSelection || hasPartialSelectionWithin) {
        return;
      }
      expressionInputSelectAll();
    },
    [expressionInputSelectAll]
  );
  const expressionOnBlurHandler = (0, import_react.useCallback)(() => {
    if (expressionInputRef.current && value?.type === "expression") {
      const currentExpressionInput = expressionInputRef.current.textContent ?? "";
      const isValidExpression = (0, import_units._parseTimeExpression)(
        currentExpressionInput.trim()
      );
      if (isValidExpression) {
        expressionInputRef.current.textContent = (0, import_timeParser.formatTimeValue)(value);
      }
    }
  }, [value]);
  const pasteText = (0, import_react.useCallback)(
    (text, skipUndoCapture = false) => {
      if (readOnly) {
        return false;
      }
      let trimmedData = (0, import_helpers2.normalizeISOInput)(text.trim());
      if (!trimmedData) {
        return true;
      }
      const partialDate = (0, import_helpers2.tryParsePartialDate)(trimmedData);
      const partialTime = (0, import_helpers2.tryParsePartialTime)(trimmedData);
      const partialResult = partialDate || partialTime ? (0, import_helpers2.partialDateTimeToISO)(partialDate, partialTime) : null;
      if (partialResult && !partialResult.partsToUpdate) {
        trimmedData = partialResult.iso;
      }
      if (!skipUndoCapture) {
        const snapshot = (0, import_helpers2.captureUndoSnapshot)(
          spinButtonHandlers,
          expressionInputRef.current,
          value,
          valueType
        );
        pasteUndoRef.current = {
          snapshot,
          focusTarget: document.activeElement,
          expressionCursorOffset: expressionInputRef.current ? (0, import_selection_helpers.getExpressionCursorOffset)(expressionInputRef.current) : void 0
        };
      }
      const dateOnly = (0, import_helpers2.parseDateOnlyToISO)(trimmedData, dateOnlyPasteAnchor);
      const timeOnly = (0, import_helpers2.parseTimeOnlyToISO)(trimmedData);
      const date = dateOnly ?? timeOnly;
      if (date) {
        setValueType("iso8601");
        let partsFilter;
        if (!spinButtonsEmpty) {
          partsFilter = dateOnly ? DATE_PART_TYPES : TIME_PART_TYPES;
        }
        applyDateTime(date, partsFilter);
        if (partsFilter && hasAnyEmptySpinButton(spinButtonHandlersRef)) {
          focusFirstUnfilledSpinButton(partsFilter);
        } else {
          onPasteComplete?.();
        }
        return true;
      }
      const nextValue = (0, import_units.parseTimeAsTimeValue)(
        trimmedData,
        import_tz.TZDateMini.tz((0, import_user_preferences.getTimezone)()).getTime(),
        precision
      );
      if (nextValue) {
        setValueType(nextValue.type);
        if (nextValue.type === "expression") {
          formatExpressionValue(nextValue);
          moveCursorToEnd(expressionInputRef.current);
          setValue(nextValue);
        } else {
          applyDateTime(nextValue.absoluteDate);
        }
        onPasteComplete?.();
        return true;
      }
      if (valueType === "iso8601" && /^\d+$/.test(trimmedData)) {
        return false;
      }
      if (partialResult?.partsToUpdate) {
        const { iso, partsToUpdate } = partialResult;
        setValueType("iso8601");
        applyDateTime(iso, partsToUpdate);
        for (const [key, handler] of spinButtonHandlersRef.current) {
          if (handler && !partsToUpdate.has(key)) {
            handler.clearValue(true);
          }
        }
        if (hasAnyEmptySpinButton(spinButtonHandlersRef)) {
          requestAnimationFrame(() => {
            focusFirstUnfilledSpinButton(partsToUpdate);
          });
        } else {
          onPasteComplete?.();
        }
        return true;
      }
      if (valueType === "expression") {
        if (expressionInputRef.current) {
          pasteIntoExpressionInput(expressionInputRef, trimmedData);
        }
      } else {
        setValueType("expression");
        if (expressionInputRef.current) {
          expressionInputRef.current.textContent = trimmedData;
          expressionInputRef.current.focus();
          moveCursorToEnd(expressionInputRef.current);
          expressionInputRef.current.dispatchEvent(
            new InputEvent("input", { bubbles: true })
          );
        }
      }
      return true;
    },
    [
      readOnly,
      spinButtonHandlers,
      expressionInputRef,
      value,
      dateOnlyPasteAnchor,
      precision,
      valueType,
      setValue,
      onPasteComplete,
      formatExpressionValue,
      applyDateTime,
      spinButtonsEmpty,
      focusFirstUnfilledSpinButton
    ]
  );
  const onPaste = (0, import_react.useCallback)(
    (event) => {
      const data = event.clipboardData?.getData("text/plain");
      if (!data) {
        event.preventDefault();
        return;
      }
      if (pasteText(data)) {
        event.preventDefault();
      }
    },
    [pasteText]
  );
  const onPasteUndoKeyDown = (0, import_react.useCallback)(
    (event) => {
      const isCtrlZ = (event.ctrlKey || event.metaKey) && event.key === "z";
      if (!isCtrlZ) {
        if (pasteUndoRef.current && (0, import_helpers2.isContentModifyingKey)(event)) {
          pasteUndoRef.current = null;
        }
        return;
      }
      if (!pasteUndoRef.current) {
        event.preventDefault();
        event.stopPropagation();
        return;
      }
      const { snapshot, focusTarget, expressionCursorOffset } = pasteUndoRef.current;
      pasteUndoRef.current = null;
      event.preventDefault();
      event.stopPropagation();
      setValueType(snapshot.valueType);
      if (snapshot.valueType === "iso8601") {
        snapshot.spinButtonTexts.forEach((text, type2) => {
          const handler = spinButtonHandlers.get(type2) ?? null;
          if (!handler) {
            return;
          }
          if (text === "") {
            handler.clearValue(true);
          } else if (handler.type === "number") {
            handler.applyText(text, true);
          } else {
            handler.searchValue(text[0] ?? "", true);
          }
        });
      } else {
        if (expressionInputRef.current) {
          expressionInputRef.current.textContent = snapshot.expressionText;
        }
        setExpressionEmpty(snapshot.expressionText === "");
      }
      setValue(snapshot.value);
      setInputValue(snapshot.value?.value ?? "");
      requestAnimationFrame(() => {
        if (focusTarget instanceof HTMLElement && focusTarget !== document.activeElement) {
          focusTarget.focus();
        }
        if (snapshot.valueType === "expression" && expressionCursorOffset !== void 0 && expressionInputRef.current) {
          (0, import_selection_helpers.restoreExpressionCursorOffset)(
            expressionInputRef.current,
            expressionCursorOffset
          );
        }
      });
    },
    [
      spinButtonHandlers,
      expressionInputRef,
      setValue,
      setValueType,
      setExpressionEmpty,
      setInputValue
    ]
  );
  const expressionOnKeyDownHandler = (0, import_react.useCallback)(
    (event) => {
      if (event.key === "Enter") {
        event.preventDefault();
        return;
      }
      if (onLeadingEdge && event.key === "Backspace" && (expressionInputRef.current?.textContent ?? "") === "") {
        event.preventDefault();
        requestAnimationFrame(() => onLeadingEdge?.());
        return;
      }
      if (onTrailingEdge && event.key === "Delete" && (expressionInputRef.current?.textContent ?? "") === "") {
        event.preventDefault();
        requestAnimationFrame(() => onTrailingEdge?.());
        return;
      }
      if (event.key === "ArrowLeft" && onLeadingEdge) {
        const selection = window.getSelection();
        if (selection?.anchorOffset === 0 && selection?.focusOffset === 0) {
          onLeadingEdge();
          event.preventDefault();
          return;
        }
      }
      if (event.key === "ArrowRight") {
        const selection = window.getSelection();
        const textContent = expressionInputRef.current?.textContent ?? "";
        if (selection?.anchorOffset === textContent.length) {
          onTrailingEdge?.();
          event.preventDefault();
          return;
        }
      }
      if ((event.key === "ArrowUp" || event.key === "ArrowDown") && !disabled && !readOnly) {
        const element = expressionInputRef.current;
        if (!element) {
          return;
        }
        const currentText = element.textContent ?? "";
        if (!currentText.trim()) {
          return;
        }
        const direction = event.key === "ArrowUp" ? 1 : -1;
        const newText = (0, import_useExpressionSpin.spinExpressionValue)(currentText, direction);
        if (newText === null) {
          return;
        }
        event.preventDefault();
        const selection = window.getSelection();
        const cursorOffset = selection?.focusOffset ?? 0;
        element.textContent = newText;
        requestAnimationFrame(() => {
          const textNode = element.firstChild;
          if (textNode && selection) {
            const newOffset = Math.min(cursorOffset, newText.length);
            const range = document.createRange();
            range.setStart(textNode, newOffset);
            range.collapse(true);
            selection.removeAllRanges();
            selection.addRange(range);
          }
        });
        element.dispatchEvent(new Event("input", { bubbles: true }));
      }
    },
    [onLeadingEdge, expressionInputRef, disabled, readOnly, onTrailingEdge]
  );
  const focus = (0, import_react.useCallback)(() => {
    switch (valueType) {
      case "expression":
        expressionInputRef.current?.focus();
        break;
      case "iso8601":
        firstSpinButtonHandler?.element?.current?.focus();
        break;
    }
  }, [valueType, firstSpinButtonHandler]);
  const switchToExpressionMode = (0, import_react.useCallback)(
    (text) => {
      const newValue = text !== "" ? (0, import_timeParser.getTimeValue)(text) ?? null : null;
      setInputValue(text);
      if (valueType === "expression") {
        setValue(newValue);
        setExpressionEmpty(text === "");
        if (expressionInputRef.current) {
          expressionInputRef.current.textContent = text;
        }
        return;
      }
      setValueType("expression");
      isSwitchingToExpressionRef.current = true;
      requestAnimationFrame(() => {
        setValue(newValue);
        setExpressionEmpty(text === "");
        if (expressionInputRef.current) {
          expressionInputRef.current.textContent = text;
        }
        isSwitchingToExpressionRef.current = false;
      });
    },
    [expressionInputRef, setValue, valueType]
  );
  (0, import_react.useEffect)(() => {
    if (isControlled || value?.type !== "expression") {
      return;
    }
    formatExpressionValue(value);
  }, []);
  (0, import_react.useEffect)(() => {
    if (firstRun.current) {
      firstRun.current = false;
      return;
    }
    if (valueType === "iso8601" && spinButtonsEmpty) {
      resetToExpressionMode();
      setTimeout(() => {
        expressionInputRef?.current?.focus();
      });
    }
  }, [spinButtonsEmpty]);
  (0, import_react.useEffect)(() => {
    if (valueType !== "expression") {
      return;
    }
    if (empty) {
      setInputValue("");
      return;
    }
    if (expressionInputRef.current?.textContent) {
      setInputValue(expressionInputRef.current?.textContent);
    }
  }, [empty]);
  return {
    empty,
    spinButtons,
    spinButtonHandlers,
    value,
    valueType,
    inputValue,
    expressionInputHandler,
    expressionOnFocusHandler,
    expressionOnBlurHandler,
    expressionOnKeyDownHandler,
    onPaste,
    pasteText,
    onPasteUndoKeyDown,
    clearPasteUndo: () => {
      pasteUndoRef.current = null;
    },
    focus,
    switchToExpressionMode,
    setInputValue
  };
};
