// @dynatrace/strato-components/forms/date-time-range-picker/hooks/useDateTimeRangePicker.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useDateTimeRangePicker_exports = {};
__export(useDateTimeRangePicker_exports, {
  useDateTimeRangePicker: () => useDateTimeRangePicker
});
module.exports = __toCommonJS(useDateTimeRangePicker_exports);
var import_tz = require("@date-fns/tz");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_units = require("@dynatrace-sdk/units");
var import_user_preferences = require("@dynatrace-sdk/user-preferences");
var import_helpers = require("../../../core/time/helpers.js");
var import_useTimeframeStore = require("../../../core/time/timeframeStore/useTimeframeStore.js");
var import_focusDateTimeInput = require("../../date-time-picker/utils/focusDateTimeInput.js");
var import_helpers2 = require("../helpers.js");
var import_selection_helpers = require("../selection-helpers.js");
const restoreSpinButtons = (ref, spinButtonTexts) => {
  spinButtonTexts.forEach((text, type) => {
    const handler = ref.current?.spinButtonHandlers.get(type) ?? null;
    if (!handler) {
      return;
    }
    if (text === "") {
      handler.clearValue(true);
    } else if (handler.type === "number") {
      handler.applyText(text, true);
    } else {
      handler.searchValue(text[0] ?? "", true);
    }
  });
};
const restoreSnapshot = (snapshot, ref) => {
  if (snapshot.valueType === "iso8601") {
    restoreSpinButtons(ref, snapshot.spinButtonTexts);
  } else if (ref.current?.expressionInputRef) {
    const invalidExpresssion = !snapshot.value?.absoluteDate;
    if (invalidExpresssion) {
      restoreSpinButtons(ref, snapshot.spinButtonTexts);
    }
    ref.current.expressionInputRef.textContent = snapshot.expressionText;
  }
};
const useDateTimeRangePicker = (props) => {
  const { placeholder, readOnly, dateTimeFromRef, dateTimeToRef } = props;
  const { formatMessage } = (0, import_react_intl.useIntl)();
  const value = (0, import_useTimeframeStore.useDateTimeRangePickerStore)(
    (state) => state.dateTimeRangePickerTimeframe
  );
  const setValue = (0, import_useTimeframeStore.useDateTimeRangePickerStore)(
    (state) => state.setDateTimeRangePickerTimeframe
  );
  const pasteUndoSnapshotRef = (0, import_react.useRef)(null);
  const localizedPlaceholder = (0, import_react.useMemo)(
    () => ({
      from: placeholder?.from ?? formatMessage({
        defaultMessage: "Start time",
        id: "gnKOHMjpyzk6hNcF",
        description: "Start of timeframe"
      }),
      to: placeholder?.to ?? formatMessage({
        defaultMessage: "End time",
        id: "sGxf5yRVhyC8Fpo/",
        description: "End of timeframe"
      })
    }),
    [formatMessage, placeholder?.from, placeholder?.to]
  );
  const onCopyCapture = (0, import_react.useCallback)(
    (event) => {
      event.preventDefault();
      event.stopPropagation();
      if (!(0, import_helpers.isFullTimeframe)(value)) {
        return;
      }
      navigator.clipboard.writeText(`${value.from.value}/${value.to.value}`).catch(() => {
      });
    },
    [value]
  );
  const onPasteCapture = (0, import_react.useCallback)(
    (e) => {
      if (readOnly) {
        return;
      }
      const data = e.clipboardData?.getData("text/plain");
      if (!data) {
        e.preventDefault();
        return;
      }
      if (!dateTimeFromRef.current || !dateTimeToRef.current) {
        return;
      }
      const focusTarget = document.activeElement;
      const isExpressionInputFocused = focusTarget === dateTimeFromRef.current?.expressionInputRef || focusTarget === dateTimeToRef.current?.expressionInputRef;
      const focusedExpression = isExpressionInputFocused ? focusTarget : null;
      const expressionCursorOffset = focusedExpression ? (0, import_selection_helpers.getExpressionCursorOffset)(focusedExpression) : void 0;
      const fromValueType = dateTimeFromRef.current?.element?.getAttribute("data-valuetype") ?? "expression";
      const toValueType = dateTimeToRef.current?.element?.getAttribute("data-valuetype") ?? "expression";
      const fromSnapshot = (0, import_helpers2.captureUndoSnapshot)(
        dateTimeFromRef.current.spinButtonHandlers,
        dateTimeFromRef.current.expressionInputRef,
        value?.from ?? null,
        fromValueType
      );
      const toSnapshot = (0, import_helpers2.captureUndoSnapshot)(
        dateTimeToRef.current.spinButtonHandlers,
        dateTimeToRef.current.expressionInputRef,
        value?.to ?? null,
        toValueType
      );
      const trimmed = data.trim();
      const hasArrowSep = trimmed.includes("\u2192");
      const parts = hasArrowSep ? trimmed.split("\u2192") : trimmed.split("/");
      const [nextFrom, nextTo] = parts;
      if (nextFrom && nextTo) {
        if (hasArrowSep) {
          pasteUndoSnapshotRef.current = {
            from: fromSnapshot,
            to: toSnapshot,
            focusTarget,
            expressionCursorOffset
          };
          dateTimeFromRef.current?.pasteText(nextFrom, true);
          dateTimeToRef.current?.pasteText(nextTo, true);
          e.preventDefault();
          e.stopPropagation();
          requestAnimationFrame(() => {
            (0, import_focusDateTimeInput.focusDateTimeInput)(null, dateTimeFromRef, dateTimeToRef);
          });
          return;
        }
        if (parts.length === 2) {
          const relativeDate = import_tz.TZDateMini.tz((0, import_user_preferences.getTimezone)()).getTime();
          const parsedFrom = (0, import_helpers2.parseDateOnlyToISO)(nextFrom) ?? (0, import_helpers2.parseTimeOnlyToISO)(nextFrom) ?? (0, import_helpers2.normalizeISOInput)(nextFrom);
          const parsedTo = (0, import_helpers2.parseDateOnlyToISO)(nextTo, "endOfDay") ?? (0, import_helpers2.parseTimeOnlyToISO)(nextTo) ?? (0, import_helpers2.normalizeISOInput)(nextTo);
          const validFrom = (0, import_units.parseTimeAsTimeValue)(parsedFrom, relativeDate);
          const validTo = (0, import_units.parseTimeAsTimeValue)(parsedTo, relativeDate);
          if (validFrom && validTo) {
            pasteUndoSnapshotRef.current = {
              from: fromSnapshot,
              to: toSnapshot,
              focusTarget
            };
            dateTimeFromRef.current?.pasteText(nextFrom, true);
            dateTimeToRef.current?.pasteText(nextTo, true);
            e.preventDefault();
            e.stopPropagation();
            return;
          }
        }
      }
      pasteUndoSnapshotRef.current = {
        from: fromSnapshot,
        to: toSnapshot,
        focusTarget,
        expressionCursorOffset
      };
    },
    [readOnly, value, dateTimeFromRef, dateTimeToRef]
  );
  const onUndoPasteKeyDown = (0, import_react.useCallback)(
    (event) => {
      const isCtrlZ = (event.ctrlKey || event.metaKey) && event.key === "z";
      if (isCtrlZ && pasteUndoSnapshotRef.current) {
        const snapshot = pasteUndoSnapshotRef.current;
        pasteUndoSnapshotRef.current = null;
        event.preventDefault();
        event.stopPropagation();
        restoreSnapshot(snapshot.from, dateTimeFromRef);
        restoreSnapshot(snapshot.to, dateTimeToRef);
        const fromValue = snapshot.from.value ?? void 0;
        const toValue = snapshot.to.value ?? void 0;
        setValue(
          fromValue || toValue ? { from: fromValue, to: toValue } : null
        );
        if (snapshot.focusTarget instanceof HTMLElement && snapshot.focusTarget !== document.activeElement) {
          snapshot.focusTarget.focus();
        }
        if (snapshot.expressionCursorOffset !== void 0 && snapshot.focusTarget instanceof HTMLSpanElement) {
          (0, import_selection_helpers.restoreExpressionCursorOffset)(
            snapshot.focusTarget,
            snapshot.expressionCursorOffset
          );
        }
      } else if (pasteUndoSnapshotRef.current && (0, import_helpers2.isContentModifyingKey)(event)) {
        pasteUndoSnapshotRef.current = null;
      }
    },
    [dateTimeFromRef, dateTimeToRef, setValue]
  );
  return {
    placeholder: localizedPlaceholder,
    onCopyCapture,
    onPasteCapture,
    onUndoPasteKeyDown
  };
};
