// @dynatrace/strato-components/esm/editors/dql-editor/DQLEditor.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import {
  Children,
  forwardRef,
  useCallback,
  useEffect,
  useRef,
  useState
} from "react";
import { useIntl } from "react-intl";
import * as actionsMenuStyles from "./actions-menu/ActionsMenu.sty.js";
import {
  ActionsMenu,
  ActionsMenuInternal
} from "./actions-menu/ActionsMenu.js";
import { getDqlConfig } from "./dql/dql-language.js";
import { useValidationCallbacks } from "./useValidationCallbacks.js";
import { useValidityInfo } from "./useValidityInfo.js";
import { useAriaLabelingProps } from "../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useSafeForwardProps } from "../../core/hooks/useSafeForwardProps.js";
import { isSlot } from "../../core/utils/is-slot.js";
import { BaseCodeEditor } from "../base-code-editor/BaseCodeEditor.js";
const COMPONENT_NAME = "DQLEditor";
const DQLEditorFwd = /* @__PURE__ */ forwardRef(
  (props, forwardedRef) => {
    const intl = useIntl();
    const defaultPlaceholder = intl.formatMessage({
      defaultMessage: "Add DQL",
      id: "Ir0YCuBX6vOM4t0J",
      description: "Default placeholder for the DQLEditor."
    });
    const {
      id,
      value = "",
      controlState,
      onChange,
      folding,
      onFoldingChange,
      defaultFolding,
      fullHeight = false,
      lineWrap = false,
      placeholder = defaultPlaceholder,
      readOnly = false,
      spellCheck = false,
      onValidationStart,
      onValidationEnd,
      onValidityChange,
      onFocus,
      onBlur,
      size = "default",
      required,
      children,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
    const ariaLabelingProps = useAriaLabelingProps(restProps);
    const forwardProps = useSafeForwardProps(restProps, ariaLabelingProps);
    const internalValueRef = useRef(value);
    useEffect(() => {
      internalValueRef.current = value;
    }, [value]);
    const handleChange = useCallback(
      (newValue) => {
        internalValueRef.current = newValue;
        onChange?.(newValue);
      },
      [onChange]
    );
    const getInternalValue = useCallback(() => internalValueRef.current, []);
    const [editorConfig, setEditorConfig] = useState();
    const [fireValidationStart, fireValidationEnd] = useValidationCallbacks(
      onValidationStart,
      onValidationEnd
    );
    const setValidityInfo = useValidityInfo(onValidityChange);
    useEffect(() => {
      setEditorConfig(
        getDqlConfig(
          readOnly,
          fireValidationStart,
          fireValidationEnd,
          setValidityInfo
        )
      );
    }, [readOnly, setValidityInfo, fireValidationStart, fireValidationEnd]);
    let hasActionsMenu = false;
    Children.forEach(children, (child) => {
      if (isSlot(child, ActionsMenu)) {
        hasActionsMenu = true;
      }
    });
    const editorClassName = hasActionsMenu ? clsx(
      consumerClassName,
      // Apply right padding on .cm-editor to prevent overlap with the menu trigger
      actionsMenuStyles.editorWithActionsMenu({ size })
    ) : consumerClassName;
    const editor = /* @__PURE__ */ jsx(
      BaseCodeEditor,
      {
        id,
        ref: forwardedRef,
        onChange: handleChange,
        placeholder,
        controlState,
        readOnly,
        defaultFolding,
        lineWrap,
        value,
        spellCheck,
        fullHeight,
        folding,
        onFoldingChange,
        configuration: editorConfig,
        onFocus,
        onBlur,
        size,
        required,
        className: editorClassName,
        style: consumerStyle,
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        ...ariaLabelingProps,
        ...behavioralTrackingProps,
        ...forwardProps,
        "data-dt-component": COMPONENT_NAME
      }
    );
    if (hasActionsMenu) {
      return /* @__PURE__ */ jsxs("div", { className: actionsMenuStyles.editorWrapper({ fullHeight }), children: [
        editor,
        /* @__PURE__ */ jsx(ActionsMenuInternal, { getValue: getInternalValue, size })
      ] });
    }
    return editor;
  }
);
const DQLEditor = Object.assign(DQLEditorFwd, {
  ActionsMenu
});
export {
  DQLEditor
};
