// @dynatrace/strato-components/charts/core/components/interactions-overlay/interactions/zoom/dragInteractions.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var dragInteractions_exports = {};
__export(dragInteractions_exports, {
  zoomDragEnd: () => zoomDragEnd,
  zoomDragMove: () => zoomDragMove,
  zoomDragStart: () => zoomDragStart
});
module.exports = __toCommonJS(dragInteractions_exports);
var import_center_position_vertically = require("../../utils/center-position-vertically.js");
var import_normalize_selection_area = require("../../utils/normalize-selection-area.js");
const zoomDragStart = ({
  event,
  onSelectStart
}) => {
  onSelectStart?.({
    triggeredBy: "pointer",
    position: event.startHoldingAt
  });
};
const zoomDragMove = ({
  event,
  height,
  width,
  onSelectChange
}) => {
  const {
    startHoldingAt: startHoldingAtProp,
    currentPosition: currentPositionProp
  } = event;
  const startHoldingAt = (0, import_center_position_vertically.centerPositionVertically)(startHoldingAtProp, height);
  const currentPosition = (0, import_center_position_vertically.centerPositionVertically)(currentPositionProp, height);
  const [start, end] = (0, import_normalize_selection_area.normalizeSelectionArea)([
    startHoldingAt,
    currentPosition
  ]);
  if (start.relative.x < 0 || end.relative.x > width) {
    return;
  }
  onSelectChange?.({
    triggeredBy: "pointer",
    start,
    end,
    mode: "zoom"
  });
};
const zoomDragEnd = ({
  event,
  height,
  width,
  lastMousePositionRef,
  onSelectEnd,
  onZoomChange
}) => {
  const { startHoldingAt: startHoldingAtProp, releasedAt: releasedAtProp } = event;
  const startHoldingAt = (0, import_center_position_vertically.centerPositionVertically)(startHoldingAtProp, height);
  const releasedAt = (0, import_center_position_vertically.centerPositionVertically)(releasedAtProp, height);
  const [start, end] = (0, import_normalize_selection_area.normalizeSelectionArea)([startHoldingAt, releasedAt]);
  if (start.relative.x < 0 || end.relative.x > width) {
    return;
  }
  const deltaStart = start.relative.x;
  const deltaEnd = end.relative.x - width;
  onSelectEnd?.({ cancelled: true });
  onZoomChange?.({
    triggeredBy: "pointer",
    deltaStart,
    deltaEnd,
    behavior: "zoom-x"
  });
  lastMousePositionRef.current = releasedAt;
};
