// @dynatrace/strato-components/charts/core/components/interactions-overlay/interactions/zoom/keyboardInteractions.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var keyboardInteractions_exports = {};
__export(keyboardInteractions_exports, {
  zoomKeyboardDown: () => zoomKeyboardDown,
  zoomKeyboardEnter: () => zoomKeyboardEnter,
  zoomKeyboardEscape: () => zoomKeyboardEscape,
  zoomKeyboardUp: () => zoomKeyboardUp
});
module.exports = __toCommonJS(keyboardInteractions_exports);
var import_lodash_es = require("lodash");
var import_constants = require("../../constants.js");
const zoomKeyboardEscape = ({
  selectionArea,
  eventCancelled,
  onSelectEnd,
  lastMousePositionRef,
  pinned,
  onHover,
  onTogglePin
}) => {
  if (!(0, import_lodash_es.isUndefined)(selectionArea)) {
    eventCancelled.current = true;
    onSelectEnd?.({ cancelled: true });
    if (lastMousePositionRef.current) {
      onHover?.({
        triggeredBy: "pointer",
        position: lastMousePositionRef.current
      });
    }
  }
  if (pinned) {
    onTogglePin?.({
      triggeredBy: "keyboard",
      pin: false,
      position: lastMousePositionRef.current
    });
  }
};
const zoomKeyboardEnter = ({
  selectionArea,
  lastMousePositionRef,
  width,
  onSelectEnd,
  onZoomChange,
  onHover
}) => {
  if (selectionArea) {
    const deltaStart = selectionArea[0].relative.x;
    const deltaEnd = selectionArea[1].relative.x - width;
    onSelectEnd?.({ cancelled: true });
    onZoomChange?.({
      triggeredBy: "keyboard",
      deltaStart,
      deltaEnd,
      behavior: "zoom-x"
    });
    if (lastMousePositionRef.current) {
      onHover?.({
        triggeredBy: "pointer",
        position: lastMousePositionRef.current
      });
    }
  }
};
const zoomKeyboardUp = ({
  width,
  onZoomChange
}) => {
  const delta = Math.floor(width * import_constants.ZOOM_FACTOR / 2);
  onZoomChange?.({
    triggeredBy: "keyboard",
    deltaStart: delta,
    deltaEnd: -delta,
    behavior: "zoom-in"
  });
};
const zoomKeyboardDown = ({
  width,
  onZoomChange
}) => {
  const delta = Math.floor(width * import_constants.ZOOM_FACTOR / 2);
  onZoomChange?.({
    triggeredBy: "keyboard",
    deltaStart: -delta,
    deltaEnd: delta,
    behavior: "zoom-out"
  });
};
