// @dynatrace/strato-components/forms/date-time-picker/hooks/useDateTimePicker.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useDateTimePicker_exports = {};
__export(useDateTimePicker_exports, {
  useDateTimePicker: () => useDateTimePicker
});
module.exports = __toCommonJS(useDateTimePicker_exports);
var import_tz = require("@date-fns/tz");
var import_date_fns = require("date-fns");
var import_react = require("react");
var import_user_preferences = require("@dynatrace-sdk/user-preferences");
var import_useDateTimeFormat = require("./useDateTimeFormat.js");
var import_useSpinButtons = require("./useSpinButtons.js");
var import_useAriaLabelingProps = require("../../../core/hooks/useAriaLabelingProps.js");
var import_usePrevious = require("../../../core/hooks/usePrevious.js");
var import_useFormFieldMessagesContext = require("../../forms-core/hooks/useFormFieldMessagesContext.js");
var import_timeParser = require("../../forms-core/time/timeParser.js");
var import_focusDateTimeInput = require("../utils/focusDateTimeInput.js");
function useDateTimePicker(props) {
  const {
    type,
    precision,
    min,
    max,
    disabled = false,
    readOnly = false,
    value: propValue,
    onChange,
    required,
    valueType
  } = props;
  const isDisabled = disabled || valueType !== "iso8601";
  const timezone = (0, import_user_preferences.getTimezone)();
  const { "aria-label": ariaLabelPrefix } = (0, import_useAriaLabelingProps.useAriaLabelingProps)(props);
  const value = propValue === "" ? null : propValue;
  const [isoDateTime, setIsoDateTime] = (0, import_react.useState)(
    (0, import_timeParser.getTimeValue)(value ?? void 0, precision)?.absoluteDate ?? null
  );
  const pendingIsoDateTimeRef = (0, import_react.useRef)(isoDateTime);
  const { toParts, emptyParts } = (0, import_useDateTimeFormat.useDateTimeFormat)(type, precision);
  const {
    state: { isValid: isValidFormControl }
  } = (0, import_useFormFieldMessagesContext.useFormFieldMessagesContext)();
  const { spinButtons, dateTime, updateParts, handlers, parts, toDateTime } = (0, import_useSpinButtons.useSpinButtons)(
    isoDateTime,
    type,
    precision,
    isDisabled,
    readOnly,
    ariaLabelPrefix,
    value
  );
  const prevDateTime = (0, import_usePrevious.usePrevious)(dateTime);
  const minDateTime = (0, import_react.useMemo)(() => {
    const candidate = min !== void 0 ? new import_tz.TZDateMini(min, timezone) : void 0;
    return (0, import_date_fns.isValid)(candidate) ? candidate : void 0;
  }, [min, timezone]);
  const maxDateTime = (0, import_react.useMemo)(() => {
    const candidate = max !== void 0 ? new import_tz.TZDateMini(max, timezone) : void 0;
    return (0, import_date_fns.isValid)(candidate) ? candidate : void 0;
  }, [max, timezone]);
  const applyDateTime = (0, import_react.useCallback)(
    (isoDateTime2, passFilter) => {
      if (isoDateTime2 === null) {
        updateParts(emptyParts);
        return true;
      }
      const newDate = new import_tz.TZDateMini(isoDateTime2, timezone);
      newDate.setMinutes(
        newDate.getMinutes() + (type === "date" ? newDate.getTimezoneOffset() : 0)
      );
      if (!(0, import_date_fns.isValid)(newDate)) {
        return false;
      }
      const newDateParts = toParts(newDate);
      passFilter ? updateParts(newDateParts.filter((p) => passFilter.has(p.type))) : updateParts(newDateParts);
      return true;
    },
    [emptyParts, timezone, toParts, type, updateParts]
  );
  const dateSelectorProps = (0, import_react.useMemo)(
    () => ({
      value: isoDateTime ?? import_tz.TZDateMini.tz(timezone).toISOString(),
      onChange: (newIsoString) => {
        applyDateTime(newIsoString, /* @__PURE__ */ new Set(["year", "month", "day"]));
        (0, import_focusDateTimeInput.focusElementWithRetry)(() => handlers.get("hour")?.element.current);
      },
      disabled: disabled || readOnly,
      error: false
    }),
    [isoDateTime, timezone, disabled, readOnly, applyDateTime, handlers]
  );
  const error = (0, import_react.useMemo)(() => {
    if (!isValidFormControl) {
      return true;
    }
    if (!required && isoDateTime === null) {
      return false;
    }
    return (dateTime === null || !(0, import_date_fns.isValid)(dateTime) || minDateTime && (0, import_date_fns.isBefore)(dateTime, minDateTime) || maxDateTime && (0, import_date_fns.isAfter)(dateTime, maxDateTime)) ?? false;
  }, [
    isValidFormControl,
    required,
    isoDateTime,
    dateTime,
    minDateTime,
    maxDateTime
  ]);
  const empty = (0, import_react.useMemo)(
    () => !parts.some((part) => part.type !== "literal" && part.value !== ""),
    [parts]
  );
  (0, import_react.useEffect)(() => {
    const numericDateTime = +(dateTime || 0);
    const numericPrevDateTime = +(prevDateTime || 0);
    if (prevDateTime === void 0 || Number.isNaN(numericDateTime) && Number.isNaN(numericPrevDateTime) || numericDateTime === numericPrevDateTime) {
      return;
    }
    if (dateTime === null) {
      pendingIsoDateTimeRef.current = null;
      setIsoDateTime(null);
      onChange?.(null);
      return;
    }
    if (!(0, import_date_fns.isValid)(dateTime)) {
      pendingIsoDateTimeRef.current = null;
      setIsoDateTime(null);
      onChange?.("Invalid Date");
      return;
    }
    const newIsoDateTime = dateTime.toISOString();
    if (isoDateTime === newIsoDateTime) {
      return;
    }
    pendingIsoDateTimeRef.current = newIsoDateTime;
    setIsoDateTime(newIsoDateTime);
    onChange?.((0, import_timeParser.getTimeValue)(newIsoDateTime ?? null, precision) ?? null);
  }, [dateTime, prevDateTime]);
  (0, import_react.useEffect)(() => {
    if ((0, import_timeParser.removePrecision)(
      pendingIsoDateTimeRef.current ?? "",
      type === "date" ? "day" : precision
    ) !== value) {
      applyDateTime(
        (0, import_timeParser.getTimeValue)(value ?? void 0, precision)?.absoluteDate ?? null
      );
    }
  }, [value]);
  return {
    isoDateTime,
    spinButtons,
    handlers,
    dateSelectorProps,
    error,
    empty,
    applyDateTime,
    firstSpinButtonHandler: handlers.get(parts[0]?.type),
    dateTime,
    toDateTime,
    parts
  };
}
