// @dynatrace/strato-components/forms/date-time-picker/hooks/useSpinButtons.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useSpinButtons_exports = {};
__export(useSpinButtons_exports, {
  useSpinButtons: () => useSpinButtons
});
module.exports = __toCommonJS(useSpinButtons_exports);
var import_jsx_runtime = (
  // Show the whole literal but only use the first char as trigger.
  require("react/jsx-runtime")
);
var import_react = require("react");
var import_tz = require("@date-fns/tz");
var import_date_fns = require("date-fns");
var import_lodash_es = require("lodash");
var import_react2 = require("react");
var import_react_intl = require("react-intl");
var import_user_preferences = require("@dynatrace-sdk/user-preferences");
var import_useDateTimeMessages = require("../../../core/time/useDateTimeMessages.js");
var import_date_time = require("../../forms-core/time/date-time.js");
var import_helpers = require("../../forms-core/time/helpers.js");
var import_timeParser = require("../../forms-core/time/timeParser.js");
var import_SpinButtonGroup = require("../../spin-buttons/spin-button-group/SpinButtonGroup.js");
var import_useDateTimeFormat = require("./useDateTimeFormat.js");
function useSpinButtons(isoDateTime, type, precision, isDisabled, isReadonly, ariaLabelPrefix, value) {
  const intl = (0, import_react_intl.useIntl)();
  const prefix = ariaLabelPrefix ? `${ariaLabelPrefix} ` : "";
  const { toParts, toDateTime, emptyParts } = (0, import_useDateTimeFormat.useDateTimeFormat)(
    type,
    precision
  );
  const timeValue = (0, import_react2.useMemo)(() => (0, import_timeParser.getTimeValue)(isoDateTime), [isoDateTime]);
  const dayPeriodValues = (0, import_react2.useMemo)(() => (0, import_helpers.getDayPeriodValues)(), []);
  const timezone = (0, import_user_preferences.getTimezone)();
  const [parts, setParts] = (0, import_react2.useState)(
    timeValue !== null && timeValue !== void 0 ? toParts(
      new import_tz.TZDateMini(
        (0, import_helpers.convertDateOnlyToDateTime)(timeValue.absoluteDate, type),
        timezone
      )
    ) : emptyParts
  );
  const hasDayPeriod = (0, import_react2.useMemo)(
    () => parts.some((p) => p.type === "dayPeriod"),
    [parts]
  );
  const precisionRef = (0, import_react2.useRef)(precision);
  const dateTime = (0, import_react2.useMemo)(() => {
    const newDateTime = toDateTime(parts);
    if (!newDateTime || !(0, import_date_fns.isValid)(newDateTime)) {
      return newDateTime;
    }
    if (precision === precisionRef.current && precision === "milliseconds") {
      return newDateTime;
    }
    const date = new import_tz.TZDateMini(
      (0, import_timeParser.getTimeValue)(value, precision)?.absoluteDate ?? (0, import_date_fns.startOfDay)(import_tz.TZDateMini.tz(timezone)).toISOString(),
      timezone
    );
    const milliseconds = date.getMilliseconds();
    const seconds = date.getUTCSeconds();
    if (precision !== precisionRef.current || precision === "minutes" || precision === void 0) {
      newDateTime.setUTCSeconds(seconds);
      precisionRef.current = precision;
    }
    return new import_tz.TZDateMini(
      (0, import_date_time.replaceSecondFractions)(
        newDateTime.toISOString(),
        milliseconds.toString()
      ),
      timezone
    );
  }, [toDateTime, parts, precision, value, timezone]);
  const daysInMonth = dateTime !== null && (0, import_date_fns.isValid)(dateTime) ? (0, import_date_fns.getDaysInMonth)(dateTime) : 31;
  const updatePart = (0, import_react2.useCallback)(
    (type2, value2) => {
      setParts((prevParts) => {
        const index = prevParts.findIndex((p) => p.type === type2);
        if (index < 0 || prevParts[index].value === value2 || type2 === void 0) {
          return prevParts;
        }
        const newParts = [...prevParts];
        newParts[index] = { type: type2, value: value2 };
        return newParts;
      });
    },
    []
  );
  const sharedProps = {
    required: false,
    disabled: isDisabled,
    readOnly: isReadonly,
    isCyclic: true
  };
  const yearProps = {
    ...sharedProps,
    displayFormat: {
      length: 4
    },
    maxValue: 9999,
    minValue: 0,
    value: parseInt(parts.find((p) => p.type === "year")?.value ?? "0"),
    onChange: (value2) => {
      updatePart("year", value2?.toString() ?? "");
    },
    "data-pattern": "yyyy"
  };
  const monthProps = {
    ...sharedProps,
    displayFormat: {
      length: 2
    },
    maxValue: 12,
    minValue: 1,
    value: parseInt(parts.find((p) => p.type === "month")?.value ?? "0"),
    onChange: (value2) => {
      const formerValue = parseInt(
        parts.find((p) => p.type === "month")?.value ?? "0"
      );
      if (value2 !== formerValue) {
        updatePart("month", value2?.toString() ?? "");
      }
    },
    "data-pattern": "MM"
  };
  const dayProps = {
    ...sharedProps,
    displayFormat: {
      length: 2
    },
    maxValue: daysInMonth,
    minValue: 1,
    value: parseInt(parts.find((p) => p.type === "day")?.value ?? "0"),
    onChange: (value2) => {
      updatePart("day", value2?.toString() ?? "");
    },
    "data-pattern": "dd"
  };
  const hourProps = {
    ...sharedProps,
    displayFormat: {
      length: 2
    },
    maxValue: hasDayPeriod ? 12 : 23,
    minValue: hasDayPeriod ? 1 : 0,
    value: parseInt(parts.find((p) => p.type === "hour")?.value ?? "0"),
    onChange: (value2) => {
      updatePart("hour", value2?.toString() ?? "");
    },
    "data-pattern": parts.find((p) => p.type === "dayPeriod") ? "hh" : "HH"
  };
  const minuteProps = {
    ...sharedProps,
    displayFormat: {
      length: 2
    },
    maxValue: 59,
    minValue: 0,
    value: parseInt(parts.find((p) => p.type === "minute")?.value ?? "0"),
    onChange: (value2) => {
      updatePart("minute", value2?.toString() ?? "");
    },
    "data-pattern": "mm"
  };
  const secondsProps = {
    ...sharedProps,
    displayFormat: {
      length: 2
    },
    maxValue: 59,
    minValue: 0,
    value: parseInt(parts.find((p) => p.type === "second")?.value ?? "0"),
    onChange: (value2) => {
      updatePart("second", value2?.toString() ?? "");
    },
    "data-pattern": "ss"
  };
  const fractionalSecondProps = {
    ...sharedProps,
    displayFormat: {
      length: 3
    },
    maxValue: 999,
    minValue: 0,
    value: parseInt(
      parts.find((p) => p.type === "fractionalSecond")?.value ?? "0"
    ),
    onChange: (value2) => {
      updatePart("fractionalSecond", value2?.toString() ?? "");
    },
    "data-pattern": "SSS"
  };
  const dayPeriodProps = {
    ...sharedProps,
    options: dayPeriodValues,
    value: parts.find((p) => p.type === "dayPeriod")?.value,
    onChange: (value2) => {
      if (value2 === null) {
        updatePart?.("dayPeriod", "");
        return;
      }
      updatePart?.("dayPeriod", value2);
    },
    "data-pattern": "aa"
  };
  const { spinButtons, handlers } = (0, import_react2.useMemo)(() => {
    const handlers2 = /* @__PURE__ */ new Map();
    const setHandler = (handler, type2) => {
      handlers2?.set?.(type2, handler);
    };
    const spinButtons2 = parts.map(({ type: type2, value: value2 }, index) => {
      switch (type2) {
        case "year":
          return /* @__PURE__ */ (0, import_react.createElement)(
            import_SpinButtonGroup.SpinButtonGroup.Number,
            {
              ...yearProps,
              key: type2,
              ref: (handler) => {
                setHandler(handler, type2);
              },
              value: yearProps.value !== void 0 ? yearProps.value : parseInt(value2),
              "aria-label": intl.formatMessage(
                {
                  defaultMessage: "{prefix}year",
                  id: "nZk5lg177+CQcO/Z",
                  description: "The aria label for the year input of a date."
                },
                {
                  prefix
                }
              ),
              placeholder: (0, import_useDateTimeMessages.getPlaceholder)(type2, intl.formatMessage)
            }
          );
        case "month":
          return /* @__PURE__ */ (0, import_react.createElement)(
            import_SpinButtonGroup.SpinButtonGroup.Number,
            {
              ...monthProps,
              key: type2,
              ref: (handler) => {
                setHandler(handler, type2);
              },
              value: monthProps.value !== void 0 ? monthProps.value : parseInt(value2),
              "aria-label": intl.formatMessage(
                {
                  defaultMessage: "{prefix}month",
                  id: "LoBSMcHvTQEABp39",
                  description: "The aria label for the month input of a date."
                },
                {
                  prefix
                }
              ),
              placeholder: (0, import_useDateTimeMessages.getPlaceholder)(type2, intl.formatMessage)
            }
          );
        case "day":
          return /* @__PURE__ */ (0, import_react.createElement)(
            import_SpinButtonGroup.SpinButtonGroup.Number,
            {
              ...dayProps,
              key: type2,
              ref: (handler) => {
                setHandler(handler, type2);
              },
              value: dayProps.value !== void 0 ? dayProps.value : parseInt(value2),
              "aria-label": intl.formatMessage(
                {
                  defaultMessage: "{prefix}day",
                  id: "Dz8K7VMW5UFpwetG",
                  description: "The aria label for the day input of a date."
                },
                {
                  prefix
                }
              ),
              placeholder: (0, import_useDateTimeMessages.getPlaceholder)(type2, intl.formatMessage)
            }
          );
        case "hour":
          return /* @__PURE__ */ (0, import_react.createElement)(
            import_SpinButtonGroup.SpinButtonGroup.Number,
            {
              ...hourProps,
              key: type2,
              ref: (handler) => {
                setHandler(handler, type2);
              },
              value: hourProps.value !== void 0 ? hourProps.value : parseInt(value2),
              "aria-label": intl.formatMessage(
                {
                  defaultMessage: "{prefix}hour",
                  id: "DXoXEb27LMuapbC4",
                  description: "The aria label for the hour input of a time."
                },
                {
                  prefix
                }
              ),
              placeholder: (0, import_useDateTimeMessages.getPlaceholder)(type2, intl.formatMessage)
            }
          );
        case "minute":
          return /* @__PURE__ */ (0, import_react.createElement)(
            import_SpinButtonGroup.SpinButtonGroup.Number,
            {
              ...minuteProps,
              key: type2,
              ref: (handler) => {
                setHandler(handler, type2);
              },
              value: minuteProps.value !== void 0 ? minuteProps.value : parseInt(value2),
              "aria-label": intl.formatMessage(
                {
                  defaultMessage: "{prefix}minute",
                  id: "pGR/6J+2uPcvCSju",
                  description: "The aria label for the minute input of a time."
                },
                {
                  prefix
                }
              ),
              placeholder: (0, import_useDateTimeMessages.getPlaceholder)(type2, intl.formatMessage)
            }
          );
        case "second":
          return /* @__PURE__ */ (0, import_react.createElement)(
            import_SpinButtonGroup.SpinButtonGroup.Number,
            {
              ...secondsProps,
              key: type2,
              ref: (handler) => {
                setHandler(handler, type2);
              },
              value: secondsProps.value !== void 0 ? secondsProps.value : parseInt(value2),
              "aria-label": intl.formatMessage(
                {
                  defaultMessage: "{prefix}second",
                  id: "WD6kswCbAgSXwMnO",
                  description: "The aria label for the second input of a time."
                },
                {
                  prefix
                }
              ),
              placeholder: (0, import_useDateTimeMessages.getPlaceholder)(type2, intl.formatMessage)
            }
          );
        case "fractionalSecond":
          return /* @__PURE__ */ (0, import_react.createElement)(
            import_SpinButtonGroup.SpinButtonGroup.Number,
            {
              ...fractionalSecondProps,
              key: type2,
              ref: (handler) => {
                setHandler(handler, type2);
              },
              value: fractionalSecondProps.value !== void 0 ? fractionalSecondProps.value : parseInt(value2),
              "aria-label": intl.formatMessage(
                {
                  defaultMessage: "{prefix}millisecond",
                  id: "wa4ct2vRPvqd0Uif",
                  description: "The aria label for the millisecond input of a time."
                },
                {
                  prefix
                }
              ),
              placeholder: (0, import_useDateTimeMessages.getPlaceholder)(type2, intl.formatMessage)
            }
          );
        case "dayPeriod":
          return /* @__PURE__ */ (0, import_react.createElement)(
            import_SpinButtonGroup.SpinButtonGroup.Text,
            {
              ...dayPeriodProps,
              key: type2,
              ref: (handler) => {
                setHandler(handler, type2);
              },
              value: dayPeriodProps.value !== void 0 ? dayPeriodProps.value : value2,
              "aria-label": intl.formatMessage(
                {
                  defaultMessage: "{prefix}day period",
                  id: "a9uVGc0cPE1OqAJj",
                  description: "The aria label for the day period (am/pm) input of a time."
                },
                {
                  prefix
                }
              ),
              placeholder: (0, import_useDateTimeMessages.getPlaceholder)(type2, intl.formatMessage)
            }
          );
        case "literal":
          return (0, import_lodash_es.isNil)(value2) ? null : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_SpinButtonGroup.SpinButtonGroup.Separator,
            {
              trigger: value2[0],
              children: value2
            },
            `${type2}_${index}`
          );
        default:
          return value2;
      }
    });
    return { handlers: handlers2, spinButtons: spinButtons2 };
  }, [
    isDisabled,
    isReadonly,
    updatePart,
    daysInMonth,
    hasDayPeriod,
    parts.length,
    yearProps.value,
    monthProps.value,
    dayProps.value,
    hourProps.value,
    minuteProps.value,
    secondsProps.value,
    fractionalSecondProps.value,
    dayPeriodProps.value
  ]);
  const updateParts = (0, import_react2.useCallback)(
    (updatedParts) => {
      const nonLiteralParts = updatedParts.filter(
        ({ type: type2 }) => type2 !== "literal"
      );
      const lookup = new Map(
        nonLiteralParts.map(({ type: type2, value: value2 }) => [type2, value2])
      );
      nonLiteralParts.forEach((part) => {
        const handler = handlers.get(part.type);
        if (part.value === "") {
          handler?.clearValue();
        } else if (handler?.type === "number") {
          handler.applyText(part.value, true);
        } else if (handler?.type === "text") {
          handler.searchValue(part.value[0], true);
        }
      });
      setParts((prevParts) => {
        return prevParts.map((part) => {
          const newValue = lookup.get(part.type);
          return newValue !== void 0 ? { ...part, value: newValue } : part;
        });
      });
    },
    [handlers]
  );
  (0, import_react2.useEffect)(() => {
    setParts(
      () => dateTime === null || !(0, import_date_fns.isValid)(dateTime) ? emptyParts : toParts(dateTime)
    );
  }, [emptyParts, toDateTime, toParts, precision]);
  return {
    spinButtons,
    dateTime,
    updateParts,
    handlers,
    parts,
    toDateTime
  };
}
