// @dynatrace/strato-components/esm/styles/getSpacingSprinkles.js
import { extract } from "./extract-util.js";
import { globalSprinkles } from "./safe-sprinkles.js";
const spacingPropKeys = [
  "padding",
  "p",
  "paddingX",
  "px",
  "paddingY",
  "py",
  "paddingTop",
  "pt",
  "paddingRight",
  "pr",
  "paddingBottom",
  "pb",
  "paddingLeft",
  "pl",
  "margin",
  "m",
  "marginX",
  "mx",
  "marginY",
  "my",
  "marginTop",
  "mt",
  "marginRight",
  "mr",
  "marginBottom",
  "mb",
  "marginLeft",
  "ml"
];
function getSpacingSprinkles(props, defaults = {}) {
  const { extracted, rest } = extract(props, spacingPropKeys);
  const spacingSprinkleConfig = {
    paddingTop: extracted.pt ?? extracted.paddingTop ?? defaults.paddingTop ?? extracted.py ?? extracted.paddingY ?? defaults.paddingY ?? extracted.p ?? extracted.padding ?? defaults.padding,
    paddingRight: extracted.pr ?? extracted.paddingRight ?? defaults.paddingRight ?? extracted.px ?? extracted.paddingX ?? defaults.paddingX ?? extracted.p ?? extracted.padding ?? defaults.padding,
    paddingBottom: extracted.pb ?? extracted.paddingBottom ?? defaults.paddingBottom ?? extracted.py ?? extracted.paddingY ?? defaults.paddingY ?? extracted.p ?? extracted.padding ?? defaults.padding,
    paddingLeft: extracted.pl ?? extracted.paddingLeft ?? defaults.paddingLeft ?? extracted.px ?? extracted.paddingX ?? defaults.paddingX ?? extracted.p ?? extracted.padding ?? defaults.padding,
    marginTop: extracted.mt ?? extracted.marginTop ?? defaults.marginTop ?? extracted.my ?? extracted.marginY ?? defaults.marginY ?? extracted.m ?? extracted.margin ?? defaults.margin,
    marginRight: extracted.mr ?? extracted.marginRight ?? defaults.marginRight ?? extracted.mx ?? extracted.marginX ?? defaults.marginX ?? extracted.m ?? extracted.margin ?? defaults.margin,
    marginBottom: extracted.mb ?? extracted.marginBottom ?? defaults.marginBottom ?? extracted.my ?? extracted.marginY ?? defaults.marginY ?? extracted.m ?? extracted.margin ?? defaults.margin,
    marginLeft: extracted.ml ?? extracted.marginLeft ?? defaults.marginLeft ?? extracted.mx ?? extracted.marginX ?? defaults.marginX ?? extracted.m ?? extracted.margin ?? defaults.margin
  };
  return [globalSprinkles(spacingSprinkleConfig), rest];
}
export {
  getSpacingSprinkles
};
