// @dynatrace/strato-components/esm/styles/index.js
import {
  getGridFlexPositionSprinkles
} from "./getGridFlexPositionSprinkles.js";
import {
  getGapSprinkles
} from "./getGapSprinkles.js";
import {
  getSpacingSprinkles
} from "./getSpacingSprinkles.js";
import {
  getLayoutSizeStyles
} from "./getLayoutSizeStyles.js";
import {
  alignContentPositionProps,
  alignItemsPositionProps,
  alignSelfPositionProps,
  justifyContentPositionProps,
  justifyItemsPositionProps,
  justifySelfPositionProps,
  marginProperties,
  placeContentPositionProps,
  placeItemsPositionProps,
  placeSelfPositionProps,
  spacingProperties
} from "./sprinkle-properties.js";
import { extract } from "./extract-util.js";
import { globalSprinkles } from "./safe-sprinkles.js";
import { textStyle } from "./textStyle.sty.js";
import { internalSprinkles } from "./sprinkles.sty.js";
export {
  extract as _extract,
  getGapSprinkles as _getGapSprinkles,
  getGridFlexPositionSprinkles as _getGridFlexPositionSprinkles,
  getLayoutSizeStyles as _getLayoutSizeStyles,
  getSpacingSprinkles as _getSpacingSprinkles,
  globalSprinkles as _globalSprinkles,
  internalSprinkles as _internalSprinkles,
  textStyle as _textStyleCSS,
  alignContentPositionProps,
  alignItemsPositionProps,
  alignSelfPositionProps,
  justifyContentPositionProps,
  justifyItemsPositionProps,
  justifySelfPositionProps,
  marginProperties,
  placeContentPositionProps,
  placeItemsPositionProps,
  placeSelfPositionProps,
  spacingProperties
};
