// @dynatrace/strato-components/esm/styles/sprinkle-properties.js
import Spacings from "@dynatrace/strato-design-tokens/spacings";
const spacingProperties = {
  0: Spacings.Size0,
  2: Spacings.Size2,
  4: Spacings.Size4,
  6: Spacings.Size6,
  8: Spacings.Size8,
  12: Spacings.Size12,
  16: Spacings.Size16,
  20: Spacings.Size20,
  24: Spacings.Size24,
  32: Spacings.Size32,
  40: Spacings.Size40,
  48: Spacings.Size48,
  56: Spacings.Size56,
  64: Spacings.Size64
};
const spacingPropertiesOrder = [
  0,
  2,
  4,
  6,
  8,
  12,
  16,
  20,
  24,
  32,
  40,
  48,
  56,
  64
];
const marginProperties = {
  ...spacingProperties,
  auto: "auto"
};
const selfPosition = {
  center: "center",
  end: "end",
  "flex-end": "flex-end",
  "flex-start": "flex-start",
  "self-end": "self-end",
  "self-start": "self-start",
  start: "start"
};
const contentDistribution = {
  "space-around": "space-around",
  "space-between": "space-between",
  "space-evenly": "space-evenly",
  stretch: "stretch"
};
const contentPosition = {
  center: "center",
  end: "end",
  "flex-end": "flex-end",
  "flex-start": "flex-start",
  start: "start"
};
const justifySelfPositionProps = {
  ...selfPosition,
  auto: "auto",
  baseline: "baseline",
  left: "left",
  normal: "normal",
  right: "right",
  stretch: "stretch"
};
const alignSelfPositionProps = {
  ...selfPosition,
  auto: "auto",
  baseline: "baseline",
  normal: "normal",
  stretch: "stretch"
};
const placeSelfPositionProps = {
  ...selfPosition,
  auto: "auto",
  baseline: "baseline",
  normal: "normal",
  stretch: "stretch"
};
const justifyItemsPositionProps = {
  ...selfPosition,
  baseline: "baseline",
  left: "left",
  legacy: "legacy",
  normal: "normal",
  right: "right",
  stretch: "stretch"
};
const alignItemsPositionProps = {
  ...selfPosition,
  baseline: "baseline",
  normal: "normal",
  stretch: "stretch"
};
const placeItemsPositionProps = {
  ...selfPosition,
  baseline: "baseline",
  normal: "normal",
  stretch: "stretch"
};
const justifyContentPositionProps = {
  ...contentDistribution,
  ...contentPosition,
  left: "left",
  normal: "normal",
  right: "right"
};
const alignContentPositionProps = {
  ...contentDistribution,
  ...contentPosition,
  baseline: "baseline",
  normal: "normal"
};
const placeContentPositionProps = {
  ...contentDistribution,
  ...contentPosition,
  baseline: "baseline",
  normal: "normal"
};
export {
  alignContentPositionProps,
  alignItemsPositionProps,
  alignSelfPositionProps,
  justifyContentPositionProps,
  justifyItemsPositionProps,
  justifySelfPositionProps,
  marginProperties,
  placeContentPositionProps,
  placeItemsPositionProps,
  placeSelfPositionProps,
  spacingProperties,
  spacingPropertiesOrder
};
