// @dynatrace/strato-components/filters/filter-field/editor-state-configs/edit-mode-update-listener.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var edit_mode_update_listener_exports = {};
__export(edit_mode_update_listener_exports, {
  handleEditMode: () => handleEditMode
});
module.exports = __toCommonJS(edit_mode_update_listener_exports);
var import_get_statement_status = require("../utils/get-statement-status.js");
var import_tree_navigation = require("../utils/tree-navigation.js");
function handleEditMode(update, suggestionsContext) {
  if (!update.focusChanged) {
    return;
  }
  const { view, state } = update;
  const cursorPos = state.selection.main.anchor;
  const statementStatus = (0, import_get_statement_status.getStatementStatus)(
    (0, import_tree_navigation.getStatementCursorAtPosition)(cursorPos, state)
  );
  const currentStatement = suggestionsContext.current?.currentStatement;
  const isAtEnd = cursorPos === state.doc.length;
  const hasTrailingSpace = state.sliceDoc(state.doc.length - 1) === " ";
  const hasTrailingBracket = state.sliceDoc(state.doc.length - 1) === ")" && currentStatement?.comparisonOperator?.type !== "InclusionOperator";
  const isStatementComplete = statementStatus === "complete";
  const shouldEnterEditMode = (
    // Focus is moved to the filter field.
    view.hasFocus && !hasTrailingSpace && !hasTrailingBracket && isStatementComplete && isAtEnd
  );
  if (!shouldEnterEditMode) {
    return;
  }
  const transaction = {
    changes: {
      from: state.doc.length,
      to: state.doc.length,
      insert: " "
    },
    selection: { anchor: state.doc.length + 1 }
  };
  view.dispatch(transaction);
}
