// @dynatrace/strato-components/filters/filter-field/editor-state-configs/editor-config.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var editor_config_exports = {};
__export(editor_config_exports, {
  ANNOTATION_COMPOSE_END: () => ANNOTATION_COMPOSE_END,
  ANNOTATION_DELETE_STATEMENT: () => ANNOTATION_DELETE_STATEMENT,
  ANNOTATION_DOC_CHANGED_CONTROLLED: () => ANNOTATION_DOC_CHANGED_CONTROLLED,
  ANNOTATION_END_PASTING_MODE: () => ANNOTATION_END_PASTING_MODE,
  ANNOTATION_SUGGESTION_APPLY: () => ANNOTATION_SUGGESTION_APPLY,
  ANNOTATION_SUGGESTION_PIN: () => ANNOTATION_SUGGESTION_PIN,
  ariaAttributesCompartment: () => ariaAttributesCompartment,
  disabledPropCompartment: () => disabledPropCompartment,
  errorsEffect: () => errorsEffect,
  overrideKeys: () => overrideKeys,
  pastingEffect: () => pastingEffect,
  pastingMode: () => pastingMode,
  pinnedFiltersEffect: () => pinnedFiltersEffect,
  placeholderCompartment: () => placeholderCompartment,
  readOnlyPropCompartment: () => readOnlyPropCompartment,
  recentFiltersEffect: () => recentFiltersEffect,
  staticEditorExtensions: () => staticEditorExtensions,
  stringAnnotation: () => stringAnnotation,
  suggestionsCompartment: () => suggestionsCompartment,
  usedErrors: () => usedErrors,
  usedPinnedFilters: () => usedPinnedFilters,
  usedRecentFilters: () => usedRecentFilters,
  usedValidatorMap: () => usedValidatorMap,
  validatorMapEffect: () => validatorMapEffect
});
module.exports = __toCommonJS(editor_config_exports);
var import_autocomplete = require("@codemirror/autocomplete");
var import_commands = require("@codemirror/commands");
var import_language = require("@codemirror/language");
var import_state = require("@codemirror/state");
var import_view = require("@codemirror/view");
var import_highlight = require("../decorations/highlight.js");
var import_filterfield_language = require("../parser/filterfield-language.js");
const placeholderCompartment = new import_state.Compartment();
const suggestionsCompartment = new import_state.Compartment();
const ariaAttributesCompartment = new import_state.Compartment();
const disabledPropCompartment = new import_state.Compartment();
const readOnlyPropCompartment = new import_state.Compartment();
const stringAnnotation = import_state.Annotation.define();
const ANNOTATION_DELETE_STATEMENT = "input.type.delete-statement";
const ANNOTATION_SUGGESTION_APPLY = "suggestion.apply";
const ANNOTATION_SUGGESTION_PIN = "suggestion.pin";
const ANNOTATION_COMPOSE_END = "input.type.compose.end";
const ANNOTATION_DOC_CHANGED_CONTROLLED = "doc-changed.controlled";
const ANNOTATION_END_PASTING_MODE = "pasting-mode.end";
const pastingEffect = import_state.StateEffect.define();
const pastingMode = import_state.StateField.define({
  create: () => {
    return { pastingInProgress: false };
  },
  update(value, transaction) {
    for (const effect of transaction.effects) {
      if (effect.is(pastingEffect)) {
        value = effect.value;
      }
    }
    return value;
  }
});
const validatorMapEffect = import_state.StateEffect.define();
const usedValidatorMap = import_state.StateField.define({
  create: () => void 0,
  update(value, transaction) {
    for (const effect of transaction.effects) {
      if (effect.is(validatorMapEffect)) {
        value = effect.value;
      }
    }
    return value;
  }
});
const errorsEffect = import_state.StateEffect.define();
const usedErrors = import_state.StateField.define({
  create() {
    return { errors: [] };
  },
  update(value, transaction) {
    for (const effect of transaction.effects) {
      if (effect.is(errorsEffect)) {
        value.errors = [];
      }
    }
    return value;
  }
});
const overrideKeys = ["ArrowUp", "ArrowDown", "Enter", "Mod-Enter"];
const staticEditorExtensions = [
  /**
   * General
   */
  // Enable line wrapping so that overflowing text is wrapped into the next line.
  import_view.EditorView.lineWrapping,
  // Enable multiple selections (CMD / ctrl + drag).
  import_state.EditorState.allowMultipleSelections.of(true),
  // Provides a undo/redo history
  (0, import_commands.history)(),
  // Draw custom cursor.
  (0, import_view.drawSelection)(),
  import_view.keymap.of([
    ...import_commands.defaultKeymap.filter(({ key }) => !overrideKeys.includes(key ?? "")),
    ...import_autocomplete.closeBracketsKeymap,
    {
      key: "Mod-z",
      run: import_commands.undo
    },
    {
      key: "Mod-y",
      run: import_commands.redo
    },
    {
      key: "Mod-Shift-z",
      run: import_commands.redo
    }
  ]),
  /**
   * Filter field language + syntax highlighting.
   *
   * Error highlighting is also done with the highlight extensions, since using an actual
   * linter, potential errors only show up after a short (configurable) delay. As soon as
   * you type something again, the errors disappear and then show up again if the input is
   * still invalid. Using the decoration plugins, we can immediately show errors.
   * The actual linter extension is only used to show tooltips with proper error messages.
   */
  // The extension for parsing the filter field language into a syntax tree.
  import_filterfield_language.filterFieldLanguageExtension,
  // Automatically close brackets.
  (0, import_autocomplete.closeBrackets)(),
  // Enable highlighting using the defined highlighter.
  // This adds the spans and classes used to style the different parts of a filter.
  (0, import_language.syntaxHighlighting)(import_highlight.filterFieldHighlighter)
];
const recentFiltersEffect = import_state.StateEffect.define();
const usedRecentFilters = import_state.StateField.define({
  create() {
    return [];
  },
  update(value, transaction) {
    for (const effect of transaction.effects) {
      if (effect.is(recentFiltersEffect)) {
        value = effect.value;
      }
    }
    return value;
  }
});
const pinnedFiltersEffect = import_state.StateEffect.define();
const usedPinnedFilters = import_state.StateField.define({
  create() {
    return [];
  },
  update(value, transaction) {
    for (const effect of transaction.effects) {
      if (effect.is(pinnedFiltersEffect)) {
        value = effect.value;
      }
    }
    return value;
  }
});
