// @dynatrace/strato-components/filters/filter-field/editor-state-configs/pasting-update-listener.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var pasting_update_listener_exports = {};
__export(pasting_update_listener_exports, {
  getChangedRange: () => getChangedRange,
  handleControlledPasting: () => handleControlledPasting,
  handlePastingTransaction: () => handlePastingTransaction
});
module.exports = __toCommonJS(pasting_update_listener_exports);
var import_editor_config = require("./editor-config.js");
function handlePastingTransaction(update, pastingInfoRef, editorView, editorViewRef) {
  const isPastingStart = update.transactions.some(
    (transaction) => transaction.isUserEvent("input.paste")
  );
  const isControlledUpdate = update.transactions.some((transaction) => {
    const annotation = transaction.annotation(import_editor_config.stringAnnotation);
    if (annotation === import_editor_config.ANNOTATION_DOC_CHANGED_CONTROLLED) {
      return true;
    }
    return false;
  });
  if (isPastingStart) {
    pastingInfoRef.current.pastingInProgress = true;
  }
  if (pastingInfoRef.current.pastingInProgress !== true) {
    return;
  }
  const isPastingEnd = update.transactions.some((transaction) => {
    return transaction.isUserEvent("input.type") || transaction.isUserEvent("delete") || transaction.isUserEvent("select");
  });
  if (isPastingEnd || update.focusChanged || pastingInfoRef.current.cancelPastingMode) {
    pastingInfoRef.current = { pastingInProgress: false };
    editorView.dispatch({
      effects: [
        import_editor_config.pastingEffect.of({
          pastingInProgress: false
        })
      ],
      // In order to re-run the decorations and re-generate errors and diagnostics,
      // we have to dispatch a new document.
      changes: {
        from: 0,
        to: editorViewRef.current?.state.doc.length,
        insert: editorViewRef.current?.state.doc
      },
      selection: update.state.selection
    });
    if (pastingInfoRef.current.cancelPastingMode) {
      pastingInfoRef.current.cancelPastingMode = false;
    }
  }
  if (update.docChanged && isPastingStart || isControlledUpdate && pastingInfoRef.current.pastingInProgress === true) {
    const changedRange = isControlledUpdate ? pastingInfoRef.current.range : getChangedRange(update);
    if (changedRange) {
      const { state } = update;
      const range = {
        from: changedRange.from,
        to: changedRange.to
      };
      editorView.dispatch({
        effects: [import_editor_config.pastingEffect.of({ pastingInProgress: true, range })],
        // In order to re-run the decorations and re-generate errors and diagnostics,
        // we have to dispatch a new document.
        changes: {
          from: 0,
          to: editorViewRef.current?.state.doc.length,
          insert: editorViewRef.current?.state.doc
        },
        selection: state.selection
      });
    }
  }
}
function getChangedRange(update) {
  const changedRanges = update.changedRanges;
  if (changedRanges && changedRanges.length > 0) {
    if (changedRanges.length === 1) {
      return { from: changedRanges[0].fromB, to: changedRanges[0].toB };
    }
    if (changedRanges.length > 1) {
      const cursorPos = update.state.selection.main.head;
      for (const range of changedRanges) {
        if (range.toB === cursorPos) {
          return { from: range.fromB, to: range.toB };
        }
      }
    }
  }
  return null;
}
function handleControlledPasting(transaction, pastingInfoRef) {
  if (transaction.isUserEvent("input.paste")) {
    const range = { from: 0, to: 0 };
    if (transaction.changes) {
      const insertions = transaction.changes.inserted;
      range.to = transaction.state.selection.main.anchor;
      if (insertions && Array.isArray(insertions) && insertions.length > 0) {
        if (insertions.length > 1) {
          range.from = Math.max(range.to - insertions[1].length, 0);
        } else {
          range.from = Math.max(range.to - insertions[0].length, 0);
        }
      } else {
        range.from = range.to;
      }
    }
    pastingInfoRef.current = { pastingInProgress: true, range };
  } else if (pastingInfoRef.current.pastingInProgress && !transaction.isUserEvent("input.paste")) {
    pastingInfoRef.current.cancelPastingMode = true;
  }
}
