// @dynatrace/strato-components/charts/single-value-grid/defaults.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var defaults_exports = {};
__export(defaults_exports, {
  DISPLAY_NAME: () => DISPLAY_NAME,
  SINGLE_VALUE_GRID_ARIA_LABEL: () => SINGLE_VALUE_GRID_ARIA_LABEL,
  gridDefaultProps: () => gridDefaultProps,
  sparklineDefaultProps: () => sparklineDefaultProps,
  trendDefaultProps: () => trendDefaultProps
});
module.exports = __toCommonJS(defaults_exports);
const DISPLAY_NAME = "SingleValueGrid";
const sparklineDefaultProps = {
  showTicks: false,
  showContextValues: false,
  variant: "line",
  curve: "linear",
  gapPolicy: "connect",
  className: "",
  style: {},
  "data-testid": "",
  "data-dtrum-mask": false,
  "data-dtrum-allow": false,
  "data-dt-name": "",
  "data-dt-content": "",
  "data-dt-children-name": "",
  "data-dt-features": "",
  "data-dt-properties": "",
  "data-dt-mouse-over": "",
  "data-dt-value-content": false,
  "data-dt-children-value-content": false,
  "data-dt-ancestor": ""
};
const trendDefaultProps = {
  showIcon: true,
  inverseTrend: "none"
};
const gridDefaultProps = {
  alignment: "start",
  applyThresholdBackground: false,
  className: "",
  style: {},
  "data-dtrum-mask": false,
  "data-dtrum-allow": false,
  "data-testid": "",
  loading: false,
  "data-dt-name": "",
  "data-dt-content": "",
  "data-dt-children-name": "",
  "data-dt-features": "",
  "data-dt-properties": "",
  "data-dt-mouse-over": "",
  "data-dt-value-content": false,
  "data-dt-children-value-content": false,
  "data-dt-ancestor": ""
};
const SINGLE_VALUE_GRID_ARIA_LABEL = "Single value data visualization grid";
