// @dynatrace/strato-components/charts/single-value-grid/SingleValueGrid.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SingleValueGrid_exports = {};
__export(SingleValueGrid_exports, {
  SingleValueGrid: () => SingleValueGrid
});
module.exports = __toCommonJS(SingleValueGrid_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_GlobalFontSizeVariables = require("./components/GlobalFontSizeVariables.js");
var import_SingleValueDisplay = require("./components/SingleValueDisplay.js");
var import_defaults = require("./defaults.js");
var import_use_label_count = require("./hooks/use-label-count.js");
var import_global_font_size_provider = require("./providers/global-font-size-provider.js");
var import_SingleValueGrid_css = require("./SingleValueGrid.sty.js");
var import_Sparkline = require("./slots/Sparkline.js");
var import_Toolbar = require("./slots/Toolbar.js");
var import_Trend = require("./slots/Trend.js");
var import_Value = require("./slots/Value.js");
var import_calculate_grid_layout = require("./utils/calculate-grid-layout.js");
var import_iterate_slots = require("./utils/iterate-slots.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_Intent = require("../../core/slots/Intent/Intent.js");
var import_Flex = require("../../layouts/flex/Flex.js");
var import_Grid = require("../../layouts/grid/Grid.js");
var import_ChartStateHandler = require("../core/components/chart-states/ChartStateHandler.js");
var import_SingleValueToolbar = require("../core/components/single-value-renderer/components/SingleValueToolbar.js");
var import_useThrottledResizeObserver = require("../core/hooks/useThrottledResizeObserver.js");
var import_ColorRule = require("../core/slots/ColorRule.js");
const COMPONENT_NAME = "SingleValueGrid";
const SingleValueGridRoot = (props) => {
  const {
    loading,
    data,
    children,
    colorAccessor,
    labelAccessor,
    prefixIconAccessor,
    applyThresholdBackground,
    style: consumerStyle,
    className: consumerClassName,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "data-testid": dataTestId,
    "aria-label": ariaLabel,
    "aria-describedby": ariaDescribedBy,
    seriesActions: individualSeriesActions,
    ...gridConfig
  } = props;
  const [behavioralTrackingProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(props);
  const labelCount = (0, import_use_label_count.useLabelCount)(data, labelAccessor);
  const { accessors: slotAccessors, ...slotsConfig } = (0, import_iterate_slots.iterateSlots)(
    children,
    props
  );
  const trendCount = slotsConfig.trend !== void 0 ? data.length : 0;
  const accessors = {
    colorAccessor,
    labelAccessor,
    prefixIconAccessor,
    ...slotAccessors
  };
  const propsConfig = {
    ...gridConfig,
    ...slotsConfig,
    applyThresholdBackground
  };
  const { height = 0, width = 0, refCallback } = (0, import_useThrottledResizeObserver.useThrottledResizeObserver)();
  const { rowCount, columnCount, lastRowColumnCount, gap } = (0, import_calculate_grid_layout.calculateGridLayout)(width, height, data.length);
  const gridItems = data.slice(0, data.length - lastRowColumnCount);
  const lastRowItems = data.slice(-lastRowColumnCount);
  const { intents, toolbar } = propsConfig;
  const displayingGridToolbar = !intents.single;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      ...behavioralTrackingProps,
      ref: refCallback,
      style: {
        ...consumerStyle
      },
      className: (0, import_clsx.default)(import_SingleValueGrid_css.gridRows),
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_ChartStateHandler.ChartStateHandler,
        {
          height,
          isLoading: !!loading,
          "data-dtrum-mask": dataDtrumMask,
          "data-dtrum-allow": dataDtrumAllow,
          style: consumerStyle,
          className: consumerClassName,
          empty: false,
          emptyState: void 0,
          errorState: void 0,
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_global_font_size_provider.GlobalFontSizeProvider,
            {
              valueCount: data.length,
              labelCount,
              trendCount,
              children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_GlobalFontSizeVariables.GlobalFontSizeVariables, { children: [
                displayingGridToolbar ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  import_SingleValueToolbar.SingleValueToolbar,
                  {
                    parentHeight: height,
                    intents: displayingGridToolbar ? intents.grid : [],
                    placement: toolbar.placement,
                    hidden: toolbar.hidden
                  }
                ) : null,
                /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
                  import_Grid.Grid,
                  {
                    className: (0, import_clsx.default)(import_SingleValueGrid_css.gridRows, consumerClassName),
                    "data-dtrum-mask": dataDtrumMask,
                    "data-dtrum-allow": dataDtrumAllow,
                    "data-testid": dataTestId,
                    "data-dt-component": COMPONENT_NAME,
                    gridTemplateRows: `repeat(${rowCount}, 1fr)`,
                    gridTemplateColumns: `repeat(${columnCount}, 1fr)`,
                    style: {
                      ...consumerStyle,
                      // Explicitly set this one via the style, as the `gridGap` settable via the
                      // props on the `Grid` component needs to be a valid spacing token
                      gridGap: gap,
                      ...displayingGridToolbar ? { pointerEvents: "none" } : {}
                    },
                    role: "img",
                    "aria-label": ariaLabel ?? import_defaults.SINGLE_VALUE_GRID_ARIA_LABEL,
                    "aria-describedby": ariaDescribedBy,
                    children: [
                      gridItems.map((record, index) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                        import_SingleValueDisplay.SingleValueDisplay,
                        {
                          record,
                          propsConfig,
                          accessors
                        },
                        index
                      )),
                      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                        import_Flex.Flex,
                        {
                          className: import_SingleValueGrid_css.gridRows,
                          style: {
                            // Explicitly set this one via the style, as the `gridGap` settable via the
                            // props on the `Grid` component needs to be a valid spacing token
                            columnGap: gap
                          },
                          gridColumn: "1 / -1",
                          children: lastRowItems.map((record, index) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                            import_SingleValueDisplay.SingleValueDisplay,
                            {
                              record,
                              propsConfig,
                              accessors
                            },
                            index
                          ))
                        }
                      )
                    ]
                  }
                )
              ] })
            }
          )
        }
      )
    }
  );
};
SingleValueGridRoot["displayName"] = import_defaults.DISPLAY_NAME;
const SingleValueGrid = Object.assign(SingleValueGridRoot, {
  ColorRule: import_ColorRule.ColorRule,
  Intent: import_Intent.Intent,
  Sparkline: import_Sparkline.Sparkline,
  Toolbar: import_Toolbar.SingleValueGridToolbarSlot,
  Trend: import_Trend.Trend,
  Value: import_Value.Value
});
