// @dynatrace/strato-components/esm/charts/xy-chart/components/displays/BarDisplay/BarDisplay.js
import { jsx } from "react/jsx-runtime";
import { BarCanvas } from "../../../../core/components/canvas/components/BarCanvas.js";
import { useGridPlotSize } from "../../../../core/components/chart-grid/hooks/useGridPlotSize.js";
import { POINT_OVERLAY_PADDING } from "../../../../core/components/data-point/constants.js";
import { DEFAULT_FILL_OPACITY } from "../../../../core/constants.js";
import { safeGet } from "../../../../core/utils/safe-get.js";
import { useBarData } from "../../../hooks/useBarData.js";
import { useXYChartData } from "../../../hooks/useXYChartData.js";
import {
  getCurrentScale,
  getDimensionScale
} from "../../../utils/get-scale.js";
import { relAccessor } from "../../../utils/value-representation.js";
import {
  heightAccessor,
  widthAccessor,
  xAccessor,
  yAccessor
} from "../bar-accessors.js";
function isXBaselineAccessor(datapoint) {
  return !!datapoint && datapoint.x0 !== void 0;
}
function isYBaselineAccessor(datapoint) {
  return !!datapoint && datapoint.y0 !== void 0;
}
function BarDisplay({
  axisIdToScales,
  series,
  metadata
}) {
  const { height, width } = useGridPlotSize();
  const xScale = getCurrentScale(
    axisIdToScales,
    series.xAxisId,
    isXBaselineAccessor(series.data[0]) ? series.type : void 0
  );
  const dimensionXScale = getDimensionScale(axisIdToScales, series.xAxisId);
  const yScale = getCurrentScale(
    axisIdToScales,
    series.yAxisId,
    isYBaselineAccessor(series.data[0]) ? series.type : void 0
  );
  const dimensionYScale = getDimensionScale(axisIdToScales, series.yAxisId);
  const { color } = metadata;
  const { datapointMetadataMap, axisIdToScaleDomain } = useXYChartData();
  const { seriesToMetadata } = useBarData();
  const { xAxis, yAxis } = safeGet(seriesToMetadata, series);
  return /* @__PURE__ */ jsx(
    BarCanvas,
    {
      data: series.data,
      canvasHeight: height,
      canvasWidth: width,
      opacity: () => DEFAULT_FILL_OPACITY,
      color: () => color.Default,
      x: xAccessor(
        xScale,
        datapointMetadataMap,
        dimensionXScale,
        series.id,
        relAccessor(xAxis, axisIdToScaleDomain)
      ),
      y: yAccessor(
        yScale,
        datapointMetadataMap,
        dimensionYScale,
        series.id,
        relAccessor(yAxis, axisIdToScaleDomain)
      ),
      width: widthAccessor(
        xScale,
        datapointMetadataMap,
        dimensionXScale,
        relAccessor(xAxis, axisIdToScaleDomain)
      ),
      height: heightAccessor(
        yScale,
        datapointMetadataMap,
        dimensionYScale,
        relAccessor(yAxis, axisIdToScaleDomain)
      ),
      enableDotOverflow: true,
      overlayPadding: POINT_OVERLAY_PADDING
    },
    series.id
  );
}
BarDisplay["displayName"] = "BarDisplay";
export {
  BarDisplay,
  isXBaselineAccessor,
  isYBaselineAccessor
};
