// @dynatrace/strato-components/esm/charts/xy-chart/components/displays/BarDisplay/BarLabelLayout.js
import { jsx } from "react/jsx-runtime";
import { BarLabelOverlay } from "./BarLabelOverlay.js";
import { plotLayout } from "../../../../core/components/chart-grid/ChartGrid.sty.js";
import { zIndicesXYChartAreas } from "../../../constants/xy-chart-z-index-areas.js";
import { useBarData } from "../../../hooks/useBarData.js";
import { useXYChartScales } from "../../../hooks/useXYChartScales.js";
function BarLabelLayout() {
  const axisIdToScales = useXYChartScales();
  const { filteredSeriesWithDatapoints, seriesToMetadata } = useBarData();
  const labelSeries = filteredSeriesWithDatapoints.filter(
    (s) => s.hasLabels === true
  );
  if (labelSeries.length === 0) {
    return null;
  }
  return /* @__PURE__ */ jsx(
    "div",
    {
      className: plotLayout,
      style: {
        zIndex: zIndicesXYChartAreas.plot + 1,
        pointerEvents: "none"
      },
      children: labelSeries.map((series) => {
        const metadata = seriesToMetadata.get(series);
        if (!metadata) {
          return null;
        }
        return /* @__PURE__ */ jsx(
          BarLabelOverlay,
          {
            axisIdToScales,
            series,
            metadata
          },
          series.id
        );
      })
    }
  );
}
BarLabelLayout["displayName"] = "BarLabelLayout";
export {
  BarLabelLayout
};
