// @dynatrace/strato-components/esm/charts/xy-chart/components/displays/BarDisplay/HorizontalBarLabel.js
import { jsx, jsxs } from "react/jsx-runtime";
import { Fragment } from "react";
import { labelClipWrapper, labelSpan } from "./BarLabelOverlay.sty.js";
import { BAR_LABEL_MIN_SIZE, BAR_LABEL_PADDING } from "./label-utils.js";
import { charsFontWidth12pixelsSize400Map } from "../../../../core/components/text-ellipsis/chars-font-width-map.js";
import { truncateString } from "../../../../core/components/text-ellipsis/truncate-string.js";
function HorizontalBarLabel({
  x,
  y,
  barWidth,
  barHeight,
  canvasWidth,
  isReversed,
  text,
  insideColor,
  outsideColor,
  truncationMode
}) {
  if (barHeight < BAR_LABEL_MIN_SIZE) {
    return null;
  }
  const centerY = y + barHeight / 2;
  const availLength = isReversed ? x + barWidth - 2 * BAR_LABEL_PADDING : canvasWidth - x - 2 * BAR_LABEL_PADDING;
  if (availLength < BAR_LABEL_MIN_SIZE) {
    return null;
  }
  const truncated = truncateString(
    text,
    availLength,
    charsFontWidth12pixelsSize400Map,
    truncationMode
  );
  const anchor = isReversed ? { right: BAR_LABEL_PADDING, textAlign: "right" } : { left: x + BAR_LABEL_PADDING };
  const outerStyle = {
    ...anchor,
    top: centerY,
    transform: "translateY(-50%)",
    width: availLength,
    color: outsideColor
  };
  const wrapperStyle = {
    left: x,
    top: y,
    width: barWidth,
    height: barHeight
  };
  const innerStyle = {
    ...isReversed ? anchor : { left: BAR_LABEL_PADDING },
    top: barHeight / 2,
    transform: "translateY(-50%)",
    width: availLength,
    color: insideColor
  };
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx("span", { className: labelSpan, style: outerStyle, children: truncated }),
    /* @__PURE__ */ jsx("div", { className: labelClipWrapper, style: wrapperStyle, children: /* @__PURE__ */ jsx("span", { className: labelSpan, style: innerStyle, children: truncated }) })
  ] });
}
HorizontalBarLabel["displayName"] = "HorizontalBarLabel";
export {
  HorizontalBarLabel
};
