// @dynatrace/strato-components/esm/charts/xy-chart/components/displays/BarDisplay/VerticalBarLabel.js
import { jsx, jsxs } from "react/jsx-runtime";
import { Fragment } from "react";
import { labelClipWrapper, labelSpan } from "./BarLabelOverlay.sty.js";
import { BAR_LABEL_MIN_SIZE, BAR_LABEL_PADDING } from "./label-utils.js";
import { charsFontWidth12pixelsSize400Map } from "../../../../core/components/text-ellipsis/chars-font-width-map.js";
import { truncateString } from "../../../../core/components/text-ellipsis/truncate-string.js";
function VerticalBarLabel({
  x,
  y,
  barWidth,
  barHeight,
  canvasHeight,
  isReversed,
  text,
  insideColor,
  outsideColor,
  truncationMode
}) {
  if (barWidth < BAR_LABEL_MIN_SIZE) {
    return null;
  }
  const availLength = canvasHeight - 2 * BAR_LABEL_PADDING;
  if (availLength < BAR_LABEL_MIN_SIZE) {
    return null;
  }
  const truncated = truncateString(
    text,
    availLength,
    charsFontWidth12pixelsSize400Map,
    truncationMode
  );
  const centerX = x + barWidth / 2;
  const centerY = canvasHeight / 2;
  const rotation = isReversed ? "rotate(90deg)" : "rotate(-90deg)";
  const outerStyle = {
    left: centerX,
    top: centerY,
    transform: `translate(-50%, -50%) ${rotation}`,
    width: availLength,
    color: outsideColor
  };
  const wrapperStyle = {
    left: x,
    top: y,
    width: barWidth,
    height: barHeight
  };
  const innerStyle = {
    left: barWidth / 2,
    top: centerY - y,
    transform: `translate(-50%, -50%) ${rotation}`,
    width: availLength,
    color: insideColor
  };
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx("span", { className: labelSpan, style: outerStyle, children: truncated }),
    /* @__PURE__ */ jsx("div", { className: labelClipWrapper, style: wrapperStyle, children: /* @__PURE__ */ jsx("span", { className: labelSpan, style: innerStyle, children: truncated }) })
  ] });
}
VerticalBarLabel["displayName"] = "VerticalBarLabel";
export {
  VerticalBarLabel
};
