// @dynatrace/strato-components/charts/xy-chart/components/tooltip/utils/get-custom-tooltip-payload.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var get_custom_tooltip_payload_exports = {};
__export(get_custom_tooltip_payload_exports, {
  getCustomTooltipPayload: () => getCustomTooltipPayload
});
module.exports = __toCommonJS(get_custom_tooltip_payload_exports);
var import_get_color = require("./get-color.js");
var import_get_tooltip_formatter = require("./get-tooltip-formatter.js");
var import_type_guards = require("../../../utils/type-guards.js");
const getCustomTooltipPayload = (content, metadata, axisIdToFormatter, intl, colorScaleMap, colorOverride) => {
  const baseTooltipPayload = {
    datapoint: content.datapoint,
    color: (0, import_get_color.getColor)(
      content.series,
      content.datapoint,
      metadata,
      colorScaleMap,
      colorOverride
    ),
    seriesId: content.series.id,
    seriesName: metadata.name,
    xFormatter: (0, import_get_tooltip_formatter.getTooltipFormatter)(metadata.xAxis, axisIdToFormatter, intl),
    yFormatter: (0, import_get_tooltip_formatter.getTooltipFormatter)(metadata.yAxis, axisIdToFormatter, intl),
    xAxisLabel: content.series.xAxisId,
    yAxisLabel: content.series.yAxisId
  };
  if ((0, import_type_guards.isRectSeries)(content.series)) {
    const rectMetadata = metadata;
    return {
      rectPayload: {
        ...baseTooltipPayload,
        datapoint: {
          ...baseTooltipPayload.datapoint,
          value: content.datapoint.value
        },
        valueLabel: rectMetadata.valueAccessorLabel ?? rectMetadata.valueAccessor ?? "",
        valueFormatter: rectMetadata.valueFormatter
      }
    };
  } else if ((0, import_type_guards.isDotSeries)(content.series)) {
    const dotMetadata = metadata;
    return {
      dotPayload: {
        ...baseTooltipPayload,
        shape: dotMetadata.shape
      }
    };
  } else if ((0, import_type_guards.isBarSeries)(content.series)) {
    return {
      barPayload: {
        ...baseTooltipPayload
      }
    };
  } else if ((0, import_type_guards.isAreaSeries)(content.series)) {
    return {
      areaPayload: {
        ...baseTooltipPayload
      }
    };
  } else {
    return {
      linePayload: {
        ...baseTooltipPayload
      }
    };
  }
};
