// @dynatrace/strato-components/charts/meter-bar/shared/components/meter-bar-layout/components/MeterBarLayoutTopRightSlot.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MeterBarLayoutTopRightSlot_exports = {};
__export(MeterBarLayoutTopRightSlot_exports, {
  MeterBarLayoutTopRightSlot: () => MeterBarLayoutTopRightSlot
});
module.exports = __toCommonJS(MeterBarLayoutTopRightSlot_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_MeterBarToolbar = require("../../meter-bar-toolbar/MeterBarToolbar.js");
var import_useComputedSegmentSize = require("../hooks/useComputedSegmentSize.js");
var import_MeterBarLayout_css = require("../MeterBarLayout.sty.js");
const toolbarWidth = "24px";
const toolbarMarginLeft = "4px";
const MeterBarLayoutTopRightSlot = (0, import_react.forwardRef)(({ isToolbarEnabled, value }, forwardedRef) => {
  const [isChartHovered, setIsChartHovered] = (0, import_react.useState)(false);
  const popoverRef = (0, import_react.useRef)(null);
  const { valueLabelRef } = (0, import_useComputedSegmentSize.useComputedSegmentSize)();
  const showPopover = () => setIsChartHovered(true);
  const hidePopover = () => setIsChartHovered(false);
  (0, import_react.useImperativeHandle)(forwardedRef, () => ({
    showPopover,
    hidePopover
  }));
  const isToolbarShown = isChartHovered && isToolbarEnabled;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    "div",
    {
      ref: valueLabelRef,
      className: (0, import_MeterBarLayout_css.valueContainer)({ inline: isToolbarEnabled }),
      children: [
        value && /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: import_MeterBarLayout_css.valueAnimatedContainer, children: value }),
        isToolbarEnabled && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          "div",
          {
            ref: popoverRef,
            className: import_MeterBarLayout_css.toolbarAnimatedContainer,
            style: {
              width: isToolbarShown ? toolbarWidth : 0,
              opacity: isToolbarShown ? 1 : 0,
              marginLeft: isToolbarShown ? toolbarMarginLeft : 0
            },
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_MeterBarToolbar.MeterBarChartToolbar, {})
          }
        )
      ]
    }
  );
});
MeterBarLayoutTopRightSlot.displayName = "MeterBarLayoutTopRightSlot";
