// @dynatrace/strato-components/layouts/page-layout/shared/portal-slots.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var portal_slots_exports = {};
__export(portal_slots_exports, {
  PageLayoutContent: () => PageLayoutContent,
  PageLayoutDetails: () => PageLayoutDetails,
  PageLayoutDetailsWithControlBar: () => PageLayoutDetailsWithControlBar,
  PageLayoutHeader: () => PageLayoutHeader,
  PageLayoutSidebar: () => PageLayoutSidebar
});
module.exports = __toCommonJS(portal_slots_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_dom = require("react-dom");
var import_PageLayoutPortalContext = require("./PageLayoutPortalContext.js");
var import_slot_defaults = require("./slot-defaults.js");
var import_useRegisteredSlot = require("./useRegisteredSlot.js");
var import_DetailsControlBar = require("../details-layout/DetailsControlBar.js");
var import_DetailsLayoutContext = require("../details-layout/DetailsLayoutContext.js");
const toRef = (ref) => ref ?? void 0;
function useShallowStable(value) {
  const ref = (0, import_react.useRef)(value);
  const prev = ref.current;
  const prevKeys = Object.keys(prev);
  const nextKeys = Object.keys(value);
  let equal = prevKeys.length === nextKeys.length;
  if (equal) {
    for (const key of nextKeys) {
      if (!Object.is(prev[key], value[key])) {
        equal = false;
        break;
      }
    }
  }
  if (!equal) {
    ref.current = value;
  }
  return ref.current;
}
const PageLayoutHeader = (0, import_react.forwardRef)(
  function PageLayoutHeader2({ children, ...domProps }, ref) {
    const { hosts } = (0, import_PageLayoutPortalContext.usePageLayoutPortalContext)();
    const stableDomProps = useShallowStable(domProps);
    const refValue = toRef(ref);
    const config = (0, import_react.useMemo)(
      () => ({ domProps: stableDomProps, ref: refValue }),
      [stableDomProps, refValue]
    );
    (0, import_useRegisteredSlot.useRegisteredSlot)("header", config);
    return (0, import_react_dom.createPortal)(children, hosts.header);
  }
);
const PageLayoutSidebar = (0, import_react.forwardRef)(
  function PageLayoutSidebar2({
    children,
    defaultWidth = import_slot_defaults.sidebarDefaults.defaultWidth,
    minWidth = import_slot_defaults.sidebarDefaults.minWidth,
    maxWidth = import_slot_defaults.sidebarDefaults.maxWidth,
    resizable = import_slot_defaults.sidebarDefaults.resizable,
    collapsed,
    defaultCollapsed = import_slot_defaults.sidebarDefaults.defaultCollapsed,
    onCollapsedChange,
    breakpoint = import_slot_defaults.sidebarDefaults.breakpoint,
    onResize,
    ...domProps
  }, ref) {
    const { hosts } = (0, import_PageLayoutPortalContext.usePageLayoutPortalContext)();
    const stableDomProps = useShallowStable(domProps);
    const refValue = toRef(ref);
    const config = (0, import_react.useMemo)(
      () => ({
        config: {
          defaultWidth,
          minWidth,
          maxWidth,
          resizable,
          collapsed,
          defaultCollapsed,
          onCollapsedChange,
          breakpoint,
          onResize
        },
        domProps: stableDomProps,
        ref: refValue
      }),
      [
        defaultWidth,
        minWidth,
        maxWidth,
        resizable,
        collapsed,
        defaultCollapsed,
        onCollapsedChange,
        breakpoint,
        onResize,
        stableDomProps,
        refValue
      ]
    );
    (0, import_useRegisteredSlot.useRegisteredSlot)("sidebar", config);
    return (0, import_react_dom.createPortal)(children, hosts.sidebar);
  }
);
const PageLayoutContent = (0, import_react.forwardRef)(
  function PageLayoutContent2({
    children,
    minWidth = import_slot_defaults.contentDefaults.minWidth,
    ...domProps
  }, ref) {
    const { hosts } = (0, import_PageLayoutPortalContext.usePageLayoutPortalContext)();
    const stableDomProps = useShallowStable(domProps);
    const refValue = toRef(ref);
    const config = (0, import_react.useMemo)(
      () => ({
        config: { minWidth },
        domProps: stableDomProps,
        ref: refValue
      }),
      [minWidth, stableDomProps, refValue]
    );
    (0, import_useRegisteredSlot.useRegisteredSlot)("content", config);
    return (0, import_react_dom.createPortal)(children, hosts.content);
  }
);
const PageLayoutDetails = (0, import_react.forwardRef)(
  function PageLayoutDetails2({
    children,
    minWidth = import_slot_defaults.detailsDefaults.minWidth,
    maxWidth = import_slot_defaults.detailsDefaults.maxWidth,
    defaultWidth = import_slot_defaults.detailsDefaults.defaultWidth,
    resizable = import_slot_defaults.detailsDefaults.resizable,
    collapsed,
    defaultCollapsed = import_slot_defaults.detailsDefaults.defaultCollapsed,
    onCollapsedChange,
    layout,
    defaultLayout = import_slot_defaults.detailsDefaults.defaultLayout,
    onLayoutChange,
    breakpoint = import_slot_defaults.detailsDefaults.breakpoint,
    onResize,
    ...domProps
  }, ref) {
    const { hosts, registry } = (0, import_PageLayoutPortalContext.usePageLayoutPortalContext)();
    const stableDomProps = useShallowStable(domProps);
    const refValue = toRef(ref);
    const config = (0, import_react.useMemo)(
      () => ({
        config: {
          minWidth,
          maxWidth,
          defaultWidth,
          resizable,
          collapsed,
          defaultCollapsed,
          onCollapsedChange,
          layout,
          defaultLayout,
          onLayoutChange,
          breakpoint,
          onResize
        },
        domProps: stableDomProps,
        ref: refValue
      }),
      [
        minWidth,
        maxWidth,
        defaultWidth,
        resizable,
        collapsed,
        defaultCollapsed,
        onCollapsedChange,
        layout,
        defaultLayout,
        onLayoutChange,
        breakpoint,
        onResize,
        stableDomProps,
        refValue
      ]
    );
    (0, import_useRegisteredSlot.useRegisteredSlot)("details", config);
    const detailsLayoutContext = (0, import_react.useSyncExternalStore)(
      registry.subscribe,
      () => registry.getSnapshot().detailsLayoutContext
    );
    return (0, import_react_dom.createPortal)(
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_DetailsLayoutContext.DetailsLayoutContext.Provider, { value: detailsLayoutContext, children }),
      hosts.details
    );
  }
);
const PageLayoutDetailsWithControlBar = Object.assign(
  PageLayoutDetails,
  { ControlBar: import_DetailsControlBar.DetailsControlBar }
);
