// @dynatrace/strato-components/layouts/page-layout/shared/slot-defaults.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var slot_defaults_exports = {};
__export(slot_defaults_exports, {
  contentDefaults: () => contentDefaults,
  detailsDefaults: () => detailsDefaults,
  sidebarDefaults: () => sidebarDefaults
});
module.exports = __toCommonJS(slot_defaults_exports);
const sidebarDefaults = {
  defaultWidth: "15%",
  minWidth: 200,
  maxWidth: "50%",
  resizable: true,
  defaultCollapsed: false,
  breakpoint: 960
};
const contentDefaults = {
  minWidth: 0
};
const detailsDefaults = {
  minWidth: 320,
  maxWidth: "50%",
  defaultWidth: "25%",
  resizable: true,
  defaultCollapsed: false,
  defaultLayout: "split",
  breakpoint: 600
};
