// @dynatrace/strato-components/layouts/page-layout/shared/slot-registry.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var slot_registry_exports = {};
__export(slot_registry_exports, {
  createSlotRegistry: () => createSlotRegistry
});
module.exports = __toCommonJS(slot_registry_exports);
const INITIAL_STATE = {
  header: null,
  sidebar: null,
  content: null,
  details: null,
  detailsLayoutContext: null
};
function createSlotRegistry() {
  let state = INITIAL_STATE;
  let frozenSnapshot = INITIAL_STATE;
  const listeners = /* @__PURE__ */ new Set();
  function notify() {
    frozenSnapshot = Object.freeze({ ...state });
    for (const listener of listeners) {
      listener();
    }
  }
  const subscribe = (listener) => {
    listeners.add(listener);
    return () => listeners.delete(listener);
  };
  const getSnapshot = () => frozenSnapshot;
  const register = (slot, config) => {
    if (state[slot] === config) {
      return;
    }
    state = { ...state, [slot]: config };
    notify();
  };
  const unregister = (slot) => {
    state = { ...state, [slot]: null };
    notify();
  };
  return { subscribe, getSnapshot, register, unregister };
}
