// @dynatrace/strato-components/layouts/page-layout/shared/useBreakpointCollapse.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useBreakpointCollapse_exports = {};
__export(useBreakpointCollapse_exports, {
  useBreakpointCollapse: () => useBreakpointCollapse
});
module.exports = __toCommonJS(useBreakpointCollapse_exports);
var import_react = require("react");
var import_use_resize_observer = __toESM(require("use-resize-observer"));
var import_useControllableState = require("../../../core/hooks/useControllableState.js");
function useBreakpointCollapse({
  collapsed: collapsedProps,
  breakpoint,
  containerRef
}) {
  const [containerWidth, setContainerWidth] = (0, import_react.useState)(0);
  const [collapsed, setCollapsed] = (0, import_useControllableState.useControllableState)({
    value: collapsedProps.value,
    defaultValue: collapsedProps.defaultValue,
    onChange: collapsedProps.onChange
  });
  const prevBelowBreakpointRef = (0, import_react.useRef)(null);
  (0, import_use_resize_observer.default)({
    ref: containerRef,
    onResize: (size) => {
      if (!size.width) {
        return;
      }
      setContainerWidth(size.width);
      const isBelowBreakpoint2 = size.width < breakpoint;
      const prevBelowBreakpoint = prevBelowBreakpointRef.current;
      prevBelowBreakpointRef.current = isBelowBreakpoint2;
      if (prevBelowBreakpoint === isBelowBreakpoint2) {
        return;
      }
      if (prevBelowBreakpoint === null && collapsed === isBelowBreakpoint2) {
        return;
      }
      setCollapsed(isBelowBreakpoint2);
    }
  });
  const isBelowBreakpoint = containerWidth !== 0 && containerWidth < breakpoint;
  return { containerWidth, collapsed, setCollapsed, isBelowBreakpoint };
}
