// @dynatrace/strato-components/charts/sparkline/Sparkline.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Sparkline_exports = {};
__export(Sparkline_exports, {
  Sparkline: () => Sparkline
});
module.exports = __toCommonJS(Sparkline_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_ThresholdIndicator = require("../core/components/sparkline-display/slots/ThresholdIndicator.js");
var import_XAxis = require("../core/components/sparkline-display/slots/XAxis.js");
var import_YAxis = require("../core/components/sparkline-display/slots/YAxis.js");
var import_SparklineDisplay = require("../core/components/sparkline-display/SparklineDisplay.js");
var import_iterate_sparkline_slots = require("../core/components/sparkline-display/utils/iterate-sparkline-slots.js");
var import_EmptyStateSlot = require("../core/slots/chart-states/EmptyStateSlot.js");
var import_ErrorStateSlot = require("../core/slots/chart-states/ErrorStateSlot.js");
const SparkLineRoot = ({
  children,
  ...props
}) => {
  const slots = (0, import_iterate_sparkline_slots.iterateSparklineSlots)(children);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_SparklineDisplay.SparklineDisplay,
    {
      ...props,
      thresholds: slots.thresholds,
      xAxis: slots.xAxis,
      yAxis: slots.yAxis,
      children
    }
  );
};
SparkLineRoot["displayName"] = "Sparkline";
const Sparkline = Object.assign(SparkLineRoot, {
  ThresholdIndicator: import_ThresholdIndicator.ThresholdIndicator,
  EmptyState: import_EmptyStateSlot.EmptyStateSlot,
  ErrorState: import_ErrorStateSlot.ErrorStateSlot,
  XAxis: import_XAxis.XAxis,
  YAxis: import_YAxis.YAxis
});
