// @dynatrace/strato-components/charts/timeseries/components/displays/bar/accessors.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var accessors_exports = {};
__export(accessors_exports, {
  getBarHeight: () => getBarHeight,
  getBarWidth: () => getBarWidth,
  getBarXPosition: () => getBarXPosition,
  getBarYPosition: () => getBarYPosition,
  height: () => height,
  width: () => width,
  x: () => x,
  y: () => y
});
module.exports = __toCommonJS(accessors_exports);
var import_detect_scale = require("../../../../core/utils/detect-scale.js");
var import_safe_get = require("../../../../core/utils/safe-get.js");
var import_yAxis = require("../../../utils/yAxis.js");
const TIMESTAMP_BAR_WIDTH = 5;
const TIMESTAMP_BAR_X_DEVIATION = 2;
const getYCoordinates = ({ value = 0, cumulativeValue = 0, unit }, scale) => {
  const yScale = (0, import_yAxis.getYScaleByUnit)(unit, scale);
  const valueCoordinate = yScale(value);
  const zeroCoordinate = yScale(0);
  const cumulativeCoordinate = yScale(cumulativeValue);
  const baseCumulative = yScale(cumulativeValue - value);
  const maxValue = yScale.invert(0);
  const maxValueCoordinate = yScale(maxValue);
  return {
    yScale,
    valueCoordinate,
    zeroCoordinate,
    cumulativeCoordinate,
    baseCumulative,
    maxValueCoordinate
  };
};
function getBarXPosition({ end, start }, xScale) {
  const isTimeframe = end && end !== start;
  return xScale(start) - (isTimeframe ? 0 : TIMESTAMP_BAR_X_DEVIATION);
}
function getBarYPosition(dataPointMetadata, yScale) {
  if (!yScale) {
    return 0;
  }
  const { valueCoordinate, zeroCoordinate, cumulativeCoordinate } = getYCoordinates(dataPointMetadata, yScale);
  let y2 = cumulativeCoordinate;
  if ((dataPointMetadata.value ?? 0) < 0) {
    const height2 = Math.abs(valueCoordinate - zeroCoordinate);
    y2 -= height2;
  }
  return y2;
}
function getBarHeight(dataPointMetadata, yScale) {
  const {
    valueCoordinate,
    zeroCoordinate,
    cumulativeCoordinate,
    baseCumulative,
    yScale: scale
  } = getYCoordinates(dataPointMetadata, yScale);
  let height2 = zeroCoordinate - valueCoordinate;
  const isLogValue = (0, import_detect_scale.isSymlogScale)(scale) && dataPointMetadata.cumulativeValue !== dataPointMetadata.value;
  if (isLogValue) {
    height2 = baseCumulative - cumulativeCoordinate;
  }
  return height2;
}
function getBarWidth({ end, start }, xScale) {
  const isTimeframe = end && end !== start;
  const width2 = isTimeframe ? xScale(end) - xScale(start) : TIMESTAMP_BAR_WIDTH;
  return width2;
}
function height(datapointsMap, yScale) {
  return (dataPoint) => {
    const dataPointMetadata = (0, import_safe_get.safeGet)(datapointsMap, dataPoint);
    return getBarHeight(dataPointMetadata, yScale);
  };
}
function width(datapointsMap, xScale) {
  return (dataPoint) => {
    const dataPointMetadata = (0, import_safe_get.safeGet)(datapointsMap, dataPoint);
    return getBarWidth(dataPointMetadata, xScale);
  };
}
function y(datapointsMap, yScale) {
  return (dataPoint) => {
    const dataPointMetadata = (0, import_safe_get.safeGet)(datapointsMap, dataPoint);
    const { cumulativeCoordinate, maxValueCoordinate } = getYCoordinates(
      dataPointMetadata,
      yScale
    );
    const yCoordinate = maxValueCoordinate + cumulativeCoordinate;
    return yCoordinate;
  };
}
function x(datapointsMap, xScale) {
  return (dataPoint) => {
    const dataPointMetadata = (0, import_safe_get.safeGet)(datapointsMap, dataPoint);
    return getBarXPosition(dataPointMetadata, xScale);
  };
}
