// @dynatrace/strato-components/charts/timeseries/components/displays/bar/BarChartDisplay.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var BarChartDisplay_exports = {};
__export(BarChartDisplay_exports, {
  BarChartDisplay: () => BarChartDisplay
});
module.exports = __toCommonJS(BarChartDisplay_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_accessors = require("./accessors.js");
var import_BarCanvas = require("../../../../core/components/canvas/components/BarCanvas.js");
var import_useGridPlotSize = require("../../../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_constants = require("../../../../core/constants.js");
var import_safe_get = require("../../../../core/utils/safe-get.js");
var import_constants2 = require("../../../constants.js");
var import_useTimeseriesFilteredData = require("../../../hooks/useTimeseriesFilteredData.js");
var import_useYScales = require("../../../hooks/useYScales.js");
const BarChartDisplay = ({
  series: barSeries,
  xScale
}) => {
  const yScale = (0, import_useYScales.useYScales)();
  const { width: widthPlot, height: heightPlot } = (0, import_useGridPlotSize.useGridPlotSize)();
  const { seriesMap, datapointsMap } = (0, import_useTimeseriesFilteredData.useTimeseriesFilteredData)();
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("figure", { "aria-label": import_constants2.BAR_CHART_DISPLAY_ARIA_LABEL, children: barSeries.map((series) => {
    const seriesMetadata = (0, import_safe_get.safeGet)(seriesMap, series);
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_BarCanvas.BarCanvas,
      {
        data: series.datapoints,
        canvasWidth: widthPlot,
        canvasHeight: heightPlot,
        x: (0, import_accessors.x)(datapointsMap, xScale),
        y: (0, import_accessors.y)(datapointsMap, yScale),
        width: (0, import_accessors.width)(datapointsMap, xScale),
        height: (0, import_accessors.height)(datapointsMap, yScale),
        strokeWidth: 1,
        color: () => seriesMetadata.color.Default,
        opacity: () => import_constants.DEFAULT_FILL_OPACITY
      },
      seriesMetadata.id
    );
  }) });
};
BarChartDisplay["displayName"] = "BarChartDisplay";
