// @dynatrace/strato-components/esm/filters/filter-field/state/recent-pinned-store.js
import { createStore } from "zustand";
import { parserVersion } from "../utils/conversion/tree-conversion.js";
function addFilterToState(storedFilters, filter, limit) {
  const trimmedFilter = filter.trim();
  const foundIndex = storedFilters.findIndex(
    (saved) => saved.filter === trimmedFilter
  );
  const newValues = [...storedFilters];
  if (foundIndex === -1) {
    newValues.push({
      filter: trimmedFilter,
      parserVersion,
      timeUsed: /* @__PURE__ */ new Date()
    });
  } else if (foundIndex > -1) {
    if (storedFilters[foundIndex].parserVersion !== parserVersion) {
      newValues[foundIndex].parserVersion = parserVersion;
    }
    storedFilters[foundIndex].timeUsed = /* @__PURE__ */ new Date();
  }
  newValues.sort((a, b) => b.timeUsed.getTime() - a.timeUsed.getTime());
  return newValues.slice(0, limit);
}
function removeFilterFromState(storedFilters, filter) {
  const trimmedFilter = filter.trim();
  const foundIndex = storedFilters.findIndex(
    (saved) => saved.filter === trimmedFilter
  );
  const newValues = [...storedFilters];
  if (foundIndex > -1) {
    newValues.splice(foundIndex, 1);
  }
  return newValues;
}
function createRecentAndPinnedFilterStore() {
  return createStore((set, get) => ({
    canPin: true,
    setCanPin: (canPin) => {
      set(() => ({ canPin }));
    },
    recentFilters: [],
    setRecentFilters: (filters) => {
      set(() => ({ recentFilters: filters }));
    },
    saveRecentFilter: (filterString) => {
      set((state) => ({
        recentFilters: addFilterToState(state.recentFilters, filterString, 100)
      }));
      return get().recentFilters;
    },
    pinnedFilters: [],
    setPinnedFilters: (filters) => {
      set(() => ({ pinnedFilters: filters }));
    },
    pinFilter: (filterString) => {
      set((state) => ({
        pinnedFilters: addFilterToState(state.pinnedFilters, filterString, 10)
      }));
      return get().pinnedFilters;
    },
    unpinFilter: (filterString) => {
      set((state) => ({
        pinnedFilters: removeFilterFromState(state.pinnedFilters, filterString)
      }));
      return get().pinnedFilters;
    }
  }));
}
const voidStore = createStore(() => ({
  canPin: true,
  setCanPin: () => void 0,
  recentFilters: [],
  setRecentFilters: () => void 0,
  saveRecentFilter: (filter) => [],
  pinnedFilters: [],
  setPinnedFilters: (filters) => void 0,
  pinFilter: (filter) => [],
  unpinFilter: (filter) => []
}));
const scopedRecentAndPinnedFiltersStores = /* @__PURE__ */ new Map([["root", createRecentAndPinnedFilterStore()]]);
export {
  addFilterToState,
  createRecentAndPinnedFilterStore,
  removeFilterFromState,
  scopedRecentAndPinnedFiltersStores,
  voidStore
};
