// @dynatrace/strato-components/esm/content/skeleton/SkeletonText.js
import { jsx, jsxs } from "react/jsx-runtime";
import { forwardRef } from "react";
import { skeletonSvgCSS } from "./Skeleton.sty.js";
import { SkeletonDrawer, SkeletonGradient } from "./utils.js";
import { useAriaLabelingProps } from "../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useId } from "../../core/hooks/useId.js";
import { useSafeForwardProps } from "../../core/hooks/useSafeForwardProps.js";
const COMPONENT_NAME = "SkeletonText";
const lineHeight = 12;
const linePadding = 8;
const baseYOffset = linePadding * 0.5;
const lineCapSize = lineHeight * 0.5;
const lineStep = lineHeight + linePadding;
const getSkeletonLinesMargin = (line, lines) => {
  if (lines === 1) {
    return 0;
  }
  return (line + 2) % 3 * lineHeight;
};
const SkeletonText = forwardRef(
  (props, forwardedRef) => {
    const {
      width = "100%",
      lines = 3,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
    const ariaLabelingProps = useAriaLabelingProps(restProps);
    const forwardProps = useSafeForwardProps(restProps, ariaLabelingProps);
    const maskId = useId();
    const gradientId = useId();
    return /* @__PURE__ */ jsx(
      "div",
      {
        ref: forwardedRef,
        className: consumerClassName,
        style: {
          width,
          height: `${lineStep * lines}px`,
          ...consumerStyle
        },
        "data-testid": dataTestId,
        ...ariaLabelingProps,
        ...behavioralTrackingProps,
        ...forwardProps,
        "data-dt-component": COMPONENT_NAME,
        children: /* @__PURE__ */ jsxs("svg", { className: skeletonSvgCSS, children: [
          /* @__PURE__ */ jsx("defs", { children: /* @__PURE__ */ jsx(SkeletonGradient, { id: gradientId }) }),
          /* @__PURE__ */ jsx("mask", { id: maskId, children: new Array(lines).fill(null).map((_, lineIndex) => /* @__PURE__ */ jsx(
            "rect",
            {
              y: baseYOffset + lineStep * lineIndex,
              height: lineHeight,
              rx: lineCapSize,
              fill: "white",
              style: {
                // we also substract 1px, to make sure the line cap is fully within the svg, and isn't cut
                width: `calc(100% - 1px - ${getSkeletonLinesMargin(
                  lineIndex,
                  lines
                )}px)`
              }
            },
            `${lines}-${lineIndex}`
          )) }),
          /* @__PURE__ */ jsx("g", { mask: `url(#${maskId})`, children: /* @__PURE__ */ jsx(SkeletonDrawer, { gradientId }) })
        ] })
      }
    );
  }
);
SkeletonText.displayName = COMPONENT_NAME;
export {
  SkeletonText
};
