// @dynatrace/strato-components/esm/content/skeleton/utils.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import TimingsJs from "@dynatrace/strato-design-tokens/timings-js";
import {
  skeletonSvgBackgroundFillCSS,
  skeletonSvgGradientCSS,
  skeletonSvgGradientRectCSS
} from "./Skeleton.sty.js";
const gradientWidth = 400;
const SkeletonGradient = (props) => /* @__PURE__ */ jsxs(
  "linearGradient",
  {
    className: skeletonSvgGradientCSS,
    id: props.id,
    x1: "0",
    x2: "1",
    children: [
      /* @__PURE__ */ jsx("stop", { offset: "0%", stopOpacity: "0.0" }),
      /* @__PURE__ */ jsx("stop", { offset: "50%", stopOpacity: "1.0" }),
      /* @__PURE__ */ jsx("stop", { offset: "100%", stopOpacity: "0.0" })
    ]
  }
);
const SkeletonBackgroundFill = () => /* @__PURE__ */ jsx("rect", { className: skeletonSvgBackgroundFillCSS });
const SkeletonDrawer = (props) => /* @__PURE__ */ jsxs(Fragment, { children: [
  /* @__PURE__ */ jsx(SkeletonBackgroundFill, {}),
  /* @__PURE__ */ jsx(
    "rect",
    {
      className: skeletonSvgGradientRectCSS,
      width: `${gradientWidth}px`,
      height: "100%",
      fill: `url(#${props.gradientId})`,
      children: /* @__PURE__ */ jsx(
        "animate",
        {
          attributeName: "x",
          from: `-${gradientWidth}px`,
          to: "100%",
          dur: `${TimingsJs.Slowest.Gentle}s`,
          repeatCount: "indefinite",
          keyTimes: "0; 1",
          calcMode: "spline",
          keySplines: "0.4, 0, 0.2, 1"
        }
      )
    }
  )
] });
export {
  SkeletonBackgroundFill,
  SkeletonDrawer,
  SkeletonGradient
};
