// @dynatrace/strato-components/styles/getSpacingSprinkles.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var getSpacingSprinkles_exports = {};
__export(getSpacingSprinkles_exports, {
  getSpacingSprinkles: () => getSpacingSprinkles
});
module.exports = __toCommonJS(getSpacingSprinkles_exports);
var import_extract_util = require("./extract-util.js");
var import_safe_sprinkles = require("./safe-sprinkles.js");
const spacingPropKeys = [
  "padding",
  "p",
  "paddingX",
  "px",
  "paddingY",
  "py",
  "paddingTop",
  "pt",
  "paddingRight",
  "pr",
  "paddingBottom",
  "pb",
  "paddingLeft",
  "pl",
  "margin",
  "m",
  "marginX",
  "mx",
  "marginY",
  "my",
  "marginTop",
  "mt",
  "marginRight",
  "mr",
  "marginBottom",
  "mb",
  "marginLeft",
  "ml"
];
function getSpacingSprinkles(props, defaults = {}) {
  const { extracted, rest } = (0, import_extract_util.extract)(props, spacingPropKeys);
  const spacingSprinkleConfig = {
    paddingTop: extracted.pt ?? extracted.paddingTop ?? defaults.paddingTop ?? extracted.py ?? extracted.paddingY ?? defaults.paddingY ?? extracted.p ?? extracted.padding ?? defaults.padding,
    paddingRight: extracted.pr ?? extracted.paddingRight ?? defaults.paddingRight ?? extracted.px ?? extracted.paddingX ?? defaults.paddingX ?? extracted.p ?? extracted.padding ?? defaults.padding,
    paddingBottom: extracted.pb ?? extracted.paddingBottom ?? defaults.paddingBottom ?? extracted.py ?? extracted.paddingY ?? defaults.paddingY ?? extracted.p ?? extracted.padding ?? defaults.padding,
    paddingLeft: extracted.pl ?? extracted.paddingLeft ?? defaults.paddingLeft ?? extracted.px ?? extracted.paddingX ?? defaults.paddingX ?? extracted.p ?? extracted.padding ?? defaults.padding,
    marginTop: extracted.mt ?? extracted.marginTop ?? defaults.marginTop ?? extracted.my ?? extracted.marginY ?? defaults.marginY ?? extracted.m ?? extracted.margin ?? defaults.margin,
    marginRight: extracted.mr ?? extracted.marginRight ?? defaults.marginRight ?? extracted.mx ?? extracted.marginX ?? defaults.marginX ?? extracted.m ?? extracted.margin ?? defaults.margin,
    marginBottom: extracted.mb ?? extracted.marginBottom ?? defaults.marginBottom ?? extracted.my ?? extracted.marginY ?? defaults.marginY ?? extracted.m ?? extracted.margin ?? defaults.margin,
    marginLeft: extracted.ml ?? extracted.marginLeft ?? defaults.marginLeft ?? extracted.mx ?? extracted.marginX ?? defaults.marginX ?? extracted.m ?? extracted.margin ?? defaults.margin
  };
  return [(0, import_safe_sprinkles.globalSprinkles)(spacingSprinkleConfig), rest];
}
