// @dynatrace/strato-components/styles/index.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var styles_exports = {};
__export(styles_exports, {
  _extract: () => import_extract_util.extract,
  _getGapSprinkles: () => import_getGapSprinkles.getGapSprinkles,
  _getGridFlexPositionSprinkles: () => import_getGridFlexPositionSprinkles.getGridFlexPositionSprinkles,
  _getLayoutSizeStyles: () => import_getLayoutSizeStyles.getLayoutSizeStyles,
  _getSpacingSprinkles: () => import_getSpacingSprinkles.getSpacingSprinkles,
  _globalSprinkles: () => import_safe_sprinkles.globalSprinkles,
  _internalSprinkles: () => import_sprinkles_css.internalSprinkles,
  _textStyleCSS: () => import_textStyle_css.textStyle,
  alignContentPositionProps: () => import_sprinkle_properties.alignContentPositionProps,
  alignItemsPositionProps: () => import_sprinkle_properties.alignItemsPositionProps,
  alignSelfPositionProps: () => import_sprinkle_properties.alignSelfPositionProps,
  justifyContentPositionProps: () => import_sprinkle_properties.justifyContentPositionProps,
  justifyItemsPositionProps: () => import_sprinkle_properties.justifyItemsPositionProps,
  justifySelfPositionProps: () => import_sprinkle_properties.justifySelfPositionProps,
  marginProperties: () => import_sprinkle_properties.marginProperties,
  placeContentPositionProps: () => import_sprinkle_properties.placeContentPositionProps,
  placeItemsPositionProps: () => import_sprinkle_properties.placeItemsPositionProps,
  placeSelfPositionProps: () => import_sprinkle_properties.placeSelfPositionProps,
  spacingProperties: () => import_sprinkle_properties.spacingProperties
});
module.exports = __toCommonJS(styles_exports);
var import_getGridFlexPositionSprinkles = require("./getGridFlexPositionSprinkles.js");
var import_getGapSprinkles = require("./getGapSprinkles.js");
var import_getSpacingSprinkles = require("./getSpacingSprinkles.js");
var import_getLayoutSizeStyles = require("./getLayoutSizeStyles.js");
var import_sprinkle_properties = require("./sprinkle-properties.js");
var import_extract_util = require("./extract-util.js");
var import_safe_sprinkles = require("./safe-sprinkles.js");
var import_textStyle_css = require("./textStyle.sty.js");
var import_sprinkles_css = require("./sprinkles.sty.js");
