// @dynatrace/strato-components/styles/sprinkle-properties.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var sprinkle_properties_exports = {};
__export(sprinkle_properties_exports, {
  alignContentPositionProps: () => alignContentPositionProps,
  alignItemsPositionProps: () => alignItemsPositionProps,
  alignSelfPositionProps: () => alignSelfPositionProps,
  justifyContentPositionProps: () => justifyContentPositionProps,
  justifyItemsPositionProps: () => justifyItemsPositionProps,
  justifySelfPositionProps: () => justifySelfPositionProps,
  marginProperties: () => marginProperties,
  placeContentPositionProps: () => placeContentPositionProps,
  placeItemsPositionProps: () => placeItemsPositionProps,
  placeSelfPositionProps: () => placeSelfPositionProps,
  spacingProperties: () => spacingProperties,
  spacingPropertiesOrder: () => spacingPropertiesOrder
});
module.exports = __toCommonJS(sprinkle_properties_exports);
var import_spacings = __toESM(require("@dynatrace/strato-design-tokens/spacings"));
const spacingProperties = {
  0: import_spacings.default.Size0,
  2: import_spacings.default.Size2,
  4: import_spacings.default.Size4,
  6: import_spacings.default.Size6,
  8: import_spacings.default.Size8,
  12: import_spacings.default.Size12,
  16: import_spacings.default.Size16,
  20: import_spacings.default.Size20,
  24: import_spacings.default.Size24,
  32: import_spacings.default.Size32,
  40: import_spacings.default.Size40,
  48: import_spacings.default.Size48,
  56: import_spacings.default.Size56,
  64: import_spacings.default.Size64
};
const spacingPropertiesOrder = [
  0,
  2,
  4,
  6,
  8,
  12,
  16,
  20,
  24,
  32,
  40,
  48,
  56,
  64
];
const marginProperties = {
  ...spacingProperties,
  auto: "auto"
};
const selfPosition = {
  center: "center",
  end: "end",
  "flex-end": "flex-end",
  "flex-start": "flex-start",
  "self-end": "self-end",
  "self-start": "self-start",
  start: "start"
};
const contentDistribution = {
  "space-around": "space-around",
  "space-between": "space-between",
  "space-evenly": "space-evenly",
  stretch: "stretch"
};
const contentPosition = {
  center: "center",
  end: "end",
  "flex-end": "flex-end",
  "flex-start": "flex-start",
  start: "start"
};
const justifySelfPositionProps = {
  ...selfPosition,
  auto: "auto",
  baseline: "baseline",
  left: "left",
  normal: "normal",
  right: "right",
  stretch: "stretch"
};
const alignSelfPositionProps = {
  ...selfPosition,
  auto: "auto",
  baseline: "baseline",
  normal: "normal",
  stretch: "stretch"
};
const placeSelfPositionProps = {
  ...selfPosition,
  auto: "auto",
  baseline: "baseline",
  normal: "normal",
  stretch: "stretch"
};
const justifyItemsPositionProps = {
  ...selfPosition,
  baseline: "baseline",
  left: "left",
  legacy: "legacy",
  normal: "normal",
  right: "right",
  stretch: "stretch"
};
const alignItemsPositionProps = {
  ...selfPosition,
  baseline: "baseline",
  normal: "normal",
  stretch: "stretch"
};
const placeItemsPositionProps = {
  ...selfPosition,
  baseline: "baseline",
  normal: "normal",
  stretch: "stretch"
};
const justifyContentPositionProps = {
  ...contentDistribution,
  ...contentPosition,
  left: "left",
  normal: "normal",
  right: "right"
};
const alignContentPositionProps = {
  ...contentDistribution,
  ...contentPosition,
  baseline: "baseline",
  normal: "normal"
};
const placeContentPositionProps = {
  ...contentDistribution,
  ...contentPosition,
  baseline: "baseline",
  normal: "normal"
};
