// @dynatrace/strato-components/charts/core/components/selection-area/Selection.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Selection_exports = {};
__export(Selection_exports, {
  Selection: () => Selection
});
module.exports = __toCommonJS(Selection_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_SelectionEdge = require("./components/SelectionEdge.js");
var import_SelectionBackdrop = require("./SelectionBackdrop.js");
var import_utils = require("./utils.js");
var import_constants = require("../overlays/constants.js");
var import_clamp_selection_area = require("./utils/clamp-selection-area.js");
var import_constants2 = require("../interactions-overlay/constants.js");
var import_normalize_selection_area = require("../interactions-overlay/utils/normalize-selection-area.js");
const extendX = (relativeAbsolute, x) => ({
  ...relativeAbsolute,
  absolute: {
    ...relativeAbsolute.absolute,
    x: relativeAbsolute.absolute.x + x
  },
  relative: {
    ...relativeAbsolute.relative,
    x: relativeAbsolute.relative.x + x
  }
});
const Selection = (props) => {
  const {
    startX,
    endX,
    height,
    width,
    focusable,
    interactable,
    bottomFormatter,
    topFormatter,
    bounds,
    handleSelectionAreaStartKeyDown,
    handleSelectionAreaEndKeyDown,
    start,
    end,
    onSelectionChange,
    isBottomLabelRotated = false,
    isTopLabelRotated = false
  } = props;
  const range = [0, width + import_constants2.SELECTION_RIGHT_OVERFLOW];
  const selectionEdgePivotRef = (0, import_react.useRef)(void 0);
  const [hoveredEdge, setHoveredEdge] = (0, import_react.useState)("");
  const onDragMove = (x) => {
    if (!(0, import_lodash_es.isUndefined)(selectionEdgePivotRef.current)) {
      const position = {
        relative: {
          x: x - bounds.left,
          y: height / 2
        },
        absolute: {
          y: height / 2 + bounds.top,
          x
        }
      };
      const nextSelection = (0, import_clamp_selection_area.clampSelectionArea)(
        (0, import_normalize_selection_area.normalizeSelectionArea)([position, selectionEdgePivotRef.current]),
        range
      );
      const edge = (0, import_utils.changedEdge)([startX, endX], nextSelection);
      if (edge) {
        setHoveredEdge(edge);
      }
      onSelectionChange?.({ selection: nextSelection, accessible: false });
    }
  };
  const onWholeSelectionMove = (x) => {
    const newSelectionAreaStart = extendX(start, x);
    const newSelectionAreaEnd = extendX(end, x);
    const nextSelection = (0, import_clamp_selection_area.clampSelectionArea)(
      (0, import_normalize_selection_area.normalizeSelectionArea)([newSelectionAreaStart, newSelectionAreaEnd]),
      range
    );
    onSelectionChange?.({ selection: nextSelection, accessible: true });
  };
  const edgesConfig = [
    {
      id: import_utils.LEFT_EDGE_ID,
      x: startX,
      onKeyDown: handleSelectionAreaStartKeyDown,
      pivot: end
    },
    {
      id: import_utils.RIGHT_EDGE_ID,
      x: endX,
      onKeyDown: handleSelectionAreaEndKeyDown,
      pivot: start
    }
  ];
  const formatters = {
    top: topFormatter || ((_x) => void 0),
    bottom: bottomFormatter || ((_x) => void 0)
  };
  const y = height / 2 - import_constants.KNOB_HEIGHT / 2;
  const pointerHanders = (id) => ({
    onPointerEnter: () => {
      setHoveredEdge(id);
    },
    onPointerLeave: () => {
      setHoveredEdge("");
    }
  });
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_SelectionBackdrop.SelectionBackdrop,
      {
        focusable,
        startX,
        endX,
        height,
        isSynchronizedSelection: false,
        interactable,
        onDragMove: onWholeSelectionMove,
        hasTopLabels: !(0, import_lodash_es.isUndefined)(topFormatter),
        hasBottomLabels: !(0, import_lodash_es.isUndefined)(bottomFormatter)
      }
    ),
    edgesConfig.map((edge) => {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_SelectionEdge.SelectionEdge,
        {
          edgeId: edge.id,
          hoveredEdge,
          x: edge.x,
          y,
          focusable,
          interactable,
          onKeyDown: edge.onKeyDown,
          onDragMove,
          onDragStart: () => {
            selectionEdgePivotRef.current = edge.pivot;
          },
          onDragEnd: () => selectionEdgePivotRef.current = void 0,
          topLabel: formatters.top(edge.x),
          bottomLabel: formatters.bottom(edge.x),
          chartBounds: bounds,
          isBottomLabelRotated,
          isTopLabelRotated,
          pointerHandlers: pointerHanders(edge.id)
        },
        edge.id
      );
    }),
    hoveredEdge === import_utils.LEFT_EDGE_ID && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_SelectionEdge.SelectionEdge,
      {
        edgeId: edgesConfig[0].id,
        hoveredEdge,
        x: edgesConfig[0].x,
        y,
        focusable,
        interactable: false,
        onKeyDown: edgesConfig[0].onKeyDown,
        onDragMove,
        onDragStart: () => {
          selectionEdgePivotRef.current = edgesConfig[0].pivot;
        },
        onDragEnd: () => selectionEdgePivotRef.current = void 0,
        topLabel: formatters.top(edgesConfig[0].x),
        bottomLabel: formatters.bottom(edgesConfig[0].x),
        chartBounds: bounds,
        isBottomLabelRotated,
        isTopLabelRotated,
        pointerHandlers: pointerHanders(edgesConfig[0].id)
      },
      edgesConfig[0].id
    )
  ] });
};
