// @dynatrace/strato-components/charts/core/components/selection-area/SelectionBackdrop.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SelectionBackdrop_exports = {};
__export(SelectionBackdrop_exports, {
  SelectionBackdrop: () => SelectionBackdrop
});
module.exports = __toCommonJS(SelectionBackdrop_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("@use-gesture/react");
var import_clsx = __toESM(require("clsx"));
var import_react2 = require("react");
var import_use_selection_area_path_ref = require("./hooks/use-selection-area-path-ref.js");
var import_SelectionArea_css = require("./SelectionArea.sty.js");
var import_constants = require("../overlays/constants.js");
const createSelectionAreaOutline = (startX, endX, height, radius, hasRoundedTop, hasRoundedBottom) => {
  const left = Math.min(startX, endX);
  const right = Math.max(startX, endX);
  const selectionWidth = right - left;
  const minimumSelectionSize = 2 * radius * 0.75;
  const canvasPixelCorrection = 0.5;
  const calculatedRadius = selectionWidth > minimumSelectionSize ? radius : Math.floor(selectionWidth * 0.75);
  const r = Math.min(calculatedRadius, (right - left) / 2, height / 2);
  let d = "";
  if (hasRoundedTop) {
    d += `M${left + r},${canvasPixelCorrection}`;
    d += `H${right - r}`;
    d += `A${r},${r} 0 0 1 ${right},${canvasPixelCorrection + r}`;
  } else {
    d += `M${left},${canvasPixelCorrection}`;
    d += `H${right}`;
  }
  if (hasRoundedBottom) {
    d += `V${height - r}`;
    d += `A${r},${r} 0 0 1 ${right - r},${height}`;
    d += `H${left + r}`;
    d += `A${r},${r} 0 0 1 ${left},${height - r}`;
  } else {
    d += `V${height}`;
    d += `H${left}`;
  }
  if (hasRoundedTop) {
    d += `V${canvasPixelCorrection + r}`;
    d += `A${r},${r} 0 0 1 ${left + r},${canvasPixelCorrection}`;
  } else {
    d += `V${canvasPixelCorrection}`;
  }
  d += "Z";
  return d;
};
const SelectionBackdrop = ({
  startX,
  endX,
  height,
  focusable,
  interactable,
  onDragMove,
  isSynchronizedSelection,
  hasTopLabels = false,
  hasBottomLabels = false
}) => {
  const selectionAreaRef = (0, import_react2.useRef)(null);
  const forwardedRef = (0, import_use_selection_area_path_ref.useSelectionAreaPathRef)();
  (0, import_react2.useImperativeHandle)(
    forwardedRef,
    () => ({
      element: selectionAreaRef.current,
      focus() {
        selectionAreaRef.current?.focus();
      }
    }),
    []
  );
  const dragHandlers = (0, import_react.useDrag)(
    ({ type, delta }) => {
      if (!(type === "pointerdown" || type === "pointerup")) {
        onDragMove?.(delta[0]);
      }
    },
    {}
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "path",
    {
      ref: selectionAreaRef,
      tabIndex: focusable ? 0 : -1,
      strokeDasharray: isSynchronizedSelection ? "3,2" : "0,0",
      className: (0, import_clsx.default)(import_SelectionArea_css.selectionArea),
      pointerEvents: interactable ? "all" : "none",
      d: createSelectionAreaOutline(
        startX,
        endX,
        height,
        import_constants.SELECTED_AREA_RADIUS,
        !hasTopLabels,
        !hasBottomLabels
      ),
      ...dragHandlers()
    }
  );
};
