// @dynatrace/strato-components/charts/core/components/selection-area/SelectionTooltipWithAsyncActions.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SelectionTooltipWithAsyncActions_exports = {};
__export(SelectionTooltipWithAsyncActions_exports, {
  SelectionTooltipWithAsyncActions: () => SelectionTooltipWithAsyncActions
});
module.exports = __toCommonJS(SelectionTooltipWithAsyncActions_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_CustomActions = require("./tooltip/CustomActions.js");
var import_sdk_logger = require("../../../../core/sdk/sdk-logger.js");
var import_Divider = require("../../../../layouts/divider/Divider.js");
var import_OverlayTooltip = require("../overlay-tooltip/OverlayTooltip.js");
var import_tooltip_css = require("../overlay-tooltip/tooltip.sty.js");
const DEBOUNCING_TIME = 150;
const emptyActionsMenu = { actions: [], intents: [] };
const SelectionTooltipWithAsyncActions = (props) => {
  const [actionsMenu, setActionsMenu] = (0, import_react.useState)(emptyActionsMenu);
  const logger = (0, import_sdk_logger.getStratoLogger)("SelectionTooltipWithAsyncActions");
  const fetchActions = (0, import_react.useCallback)(
    (0, import_lodash_es.debounce)(
      (actions) => {
        const promise = actions();
        if (promise) {
          promise.then((result) => {
            if (result.actions.length || result.intents.length) {
              setActionsMenu(result);
            }
          }).catch((error) => {
            logger.error("Failed to load select series actions", error);
          });
        }
      },
      DEBOUNCING_TIME,
      { leading: false, trailing: true }
    ),
    [logger]
  );
  (0, import_react.useEffect)(() => {
    if (!!actionsMenu.actions.length || !!actionsMenu.intents.length) {
      setActionsMenu(emptyActionsMenu);
    }
    props.actions && fetchActions(props.actions);
    return () => {
      fetchActions.cancel();
    };
  }, [props.range[0], props.range[1]]);
  const hasActions = !!actionsMenu.actions.length;
  const hasIntents = !!actionsMenu.intents.length;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_OverlayTooltip.OverlayTooltip,
    {
      isInteractive: props.selectFinished,
      absoluteX: props.position.x,
      absoluteY: props.position.y,
      children: [
        props.header,
        (!!props.children || hasActions || hasIntents) && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_Divider.Divider,
            {
              orientation: "horizontal",
              className: import_tooltip_css.divider
            },
            "headerDivider"
          ),
          /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
            "div",
            {
              style: {
                maxHeight: "96px",
                overflowY: "auto",
                overflowX: "hidden"
              },
              children: [
                props.children,
                hasActions ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_CustomActions.CustomActions, { actions: actionsMenu.actions }) : null,
                hasIntents && (hasActions || !!props.children) ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  import_Divider.Divider,
                  {
                    orientation: "horizontal",
                    className: import_tooltip_css.divider
                  },
                  "actionsDivider"
                ) : null,
                hasIntents ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_CustomActions.CustomIntents, { intents: actionsMenu?.intents }) : null
              ]
            }
          )
        ] })
      ]
    }
  );
};
