// @dynatrace/strato-components/charts/core/components/selection-area/ZoomSelection.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ZoomSelection_exports = {};
__export(ZoomSelection_exports, {
  ZoomSelection: () => ZoomSelection
});
module.exports = __toCommonJS(ZoomSelection_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_VerticalCrosshair = require("../crosshair/VerticalCrosshair.js");
var import_constants = require("../interactions-overlay/constants.js");
var import_constants2 = require("../overlays/constants.js");
var import_use_selection_area_path_ref = require("./hooks/use-selection-area-path-ref.js");
const ZoomSelection = ({
  startX,
  endX,
  height,
  focusable,
  interactable,
  bottomFormatter,
  topFormatter,
  bounds,
  handleSelectionAreaKeyDown,
  handleSelectionAreaStartKeyDown,
  handleSelectionAreaEndKeyDown,
  isBottomLabelRotated = false,
  isTopLabelRotated = false
}) => {
  const areaStyles = import_constants2.ZOOM_AREA_STYLES;
  const selectionAreaRef = (0, import_react.useRef)(null);
  const forwardedRef = (0, import_use_selection_area_path_ref.useSelectionAreaPathRef)();
  (0, import_react.useImperativeHandle)(
    forwardedRef,
    () => ({
      element: selectionAreaRef.current,
      focus() {
        selectionAreaRef.current?.focus();
      }
    }),
    []
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "rect",
      {
        ref: selectionAreaRef,
        x: startX,
        y: 0,
        width: endX - startX,
        height,
        tabIndex: focusable ? 0 : -1,
        onKeyDown: handleSelectionAreaKeyDown,
        ...areaStyles,
        rx: "0"
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "rect",
      {
        x: startX - import_constants.SELECTION_EDGE_THRESHOLD,
        y: 0,
        width: import_constants.SELECTION_EDGE_THRESHOLD * 5,
        height,
        fill: "transparent",
        onMouseDown: () => false,
        style: {
          cursor: "col-resize",
          pointerEvents: interactable ? "all" : "none"
        }
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_VerticalCrosshair.VerticalCrosshair,
      {
        x: startX,
        bottomLabel: bottomFormatter ? bottomFormatter(startX) : void 0,
        topLabel: topFormatter ? topFormatter(startX) : void 0,
        height,
        dashed: false,
        focusable,
        chartBounds: bounds,
        onKeyDown: handleSelectionAreaStartKeyDown,
        variant: "zoom",
        isTopLabelRotated,
        isBottomLabelRotated
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "rect",
      {
        x: endX - import_constants.SELECTION_EDGE_THRESHOLD,
        y: 0,
        width: import_constants.SELECTION_EDGE_THRESHOLD * 5,
        height,
        fill: "transparent",
        style: {
          cursor: "col-resize",
          pointerEvents: interactable ? "all" : "none"
        }
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_VerticalCrosshair.VerticalCrosshair,
      {
        x: endX,
        bottomLabel: bottomFormatter ? bottomFormatter(endX) : void 0,
        topLabel: topFormatter ? topFormatter(endX) : void 0,
        height,
        dashed: false,
        focusable,
        chartBounds: bounds,
        onKeyDown: handleSelectionAreaEndKeyDown,
        variant: "zoom",
        isTopLabelRotated,
        isBottomLabelRotated
      }
    )
  ] });
};
