// @dynatrace/strato-components/esm/tables/DataTable/features/UserActions/FloatingMenu.js
import { jsx, jsxs } from "react/jsx-runtime";
import { useMemo, Children } from "react";
import { createPortal } from "react-dom";
import * as styles from "./UserActions.sty.js";
import { Menu } from "../../../../navigation/menu/Menu.js";
import {
  DataTableCellContextProvider,
  DataTableColumnContextProvider,
  DataTableHeaderContextProvider,
  useDataTable
} from "../../contexts/DataTableContext.js";
function CellActionFloatingMenu() {
  const table = useDataTable();
  const { type, headerId, rowId, columnId, x, y } = table.getState()?.floatingMenuInfo ?? {
    type: null,
    headerId: null,
    rowId: null,
    columnId: null
  };
  const actions = useMemo(() => {
    if (type === "headerAction" && columnId !== null) {
      return table.getUserActionsForColumn(columnId);
    }
    if (type === "cellAction" && columnId !== null && rowId !== null) {
      return table.getUserActionsForCell(columnId, rowId);
    }
    return null;
  }, [rowId, columnId, type, table]);
  const menuOpen = type !== null && columnId !== null && actions !== null;
  const { header, column, cell } = useMemo(() => {
    if (menuOpen === false) {
      return { column: void 0, cell: void 0 };
    }
    const header2 = headerId ? table.getHeader(headerId) : void 0;
    const column2 = table.getColumn(columnId);
    let cell2 = void 0;
    if (rowId) {
      const rowCells = table.getRow(rowId).getAllCells();
      const columnCell = rowCells.find((cell3) => cell3.column.id === columnId);
      cell2 = columnCell;
    }
    return { header: header2, column: column2, cell: cell2 };
  }, [columnId, headerId, menuOpen, rowId, table]);
  if (menuOpen === false) {
    return;
  }
  return createPortal(
    /* @__PURE__ */ jsxs(
      Menu,
      {
        open: menuOpen,
        onOpenChange: (isOpen) => {
          if (isOpen === false) {
            table.setState((state) => {
              const newState = { ...state };
              newState.floatingMenuInfo = {
                type: null,
                headerId: null,
                rowId: null,
                columnId: null,
                x: 0,
                y: 0
              };
              newState.openMenuRowId = null;
              return newState;
            });
          }
        },
        modal: false,
        children: [
          /* @__PURE__ */ jsx(Menu.Trigger, { children: /* @__PURE__ */ jsx(
            "button",
            {
              disabled: true,
              type: "button",
              className: styles.floatingMenuButton,
              style: {
                top: `${y}px`,
                left: `${x}px`
              },
              "data-trigger": "context-action"
            }
          ) }),
          menuOpen === true && /* @__PURE__ */ jsx(DataTableHeaderContextProvider, { value: header, children: /* @__PURE__ */ jsx(DataTableColumnContextProvider, { value: column, children: /* @__PURE__ */ jsx(DataTableCellContextProvider, { value: cell, children: /* @__PURE__ */ jsx(Menu.Content, { children: Children.map(actions, (child) => child) }) }) }) })
        ]
      }
    ),
    document.body
  );
}
export {
  CellActionFloatingMenu
};
