// @dynatrace/strato-components/esm/tables/DataTable/features/UserActions/InlineUserActionsMenu.js
import { jsx, jsxs } from "react/jsx-runtime";
import { Children, useState } from "react";
import { DotMenuIcon } from "@dynatrace/strato-icons";
import { useCellAndColumnActions } from "./useCellActions.js";
import * as styles from "./UserActions.sty.js";
import { Menu } from "../../../../navigation/menu/Menu.js";
import {
  DATA_TABLE_CELL_HORIZONTAL_PADDING_MAP,
  DATA_TABLE_DEFAULT_ICON_SIZE,
  DATA_TABLE_SUB_ROW_EXPANDER_ICON_SIZE as SUB_ROW_EXPANDER_ICON_SIZE,
  DATA_TABLE_SUB_ROW_INDENT_GRID as SUB_ROW_INDENT_GRID
} from "../../constants.js";
import {
  DataTableColumnContextProvider,
  useDataTable
} from "../../contexts/DataTableContext.js";
function InlineUserActionsMenu(props) {
  const {
    columnId,
    headerGroupOrRowId,
    type,
    "aria-label": ariaLabel,
    "data-header-row": dataHeaderRow,
    "data-column": dataColumn,
    hovered,
    onClose
  } = props;
  const [menuOpen, setMenuOpen] = useState(false);
  const table = useDataTable();
  const column = table.getColumn(columnId);
  const actions = useCellAndColumnActions({
    enabled: menuOpen,
    headerGroupOrRowId,
    columnId,
    type
  });
  const floatingMenuInfo = table.getState().floatingMenuInfo;
  const isFloatingMenuOpenForThis = floatingMenuInfo?.columnId === columnId && (type === "cell" && floatingMenuInfo?.rowId === headerGroupOrRowId || type === "header" && floatingMenuInfo?.type === "headerAction");
  if (isFloatingMenuOpenForThis) {
    return void 0;
  }
  function getSubrowIndentation() {
    if (!table.options.enableExpanding) {
      return void 0;
    }
    if (table.getSubRowIndicatorColumnId() !== columnId) {
      return void 0;
    }
    const rowDensity = table.getVisualVariantOptions().rowDensity;
    if (type === "header") {
      if (column?.columnDef.meta?.hasCustomHeaderRenderer) {
        return void 0;
      }
      return SUB_ROW_EXPANDER_ICON_SIZE + DATA_TABLE_CELL_HORIZONTAL_PADDING_MAP[rowDensity];
    }
    const row = table.getRow(headerGroupOrRowId);
    return row.depth * SUB_ROW_INDENT_GRID + SUB_ROW_EXPANDER_ICON_SIZE + DATA_TABLE_CELL_HORIZONTAL_PADDING_MAP[rowDensity];
  }
  const triggerProps = {
    type: "button",
    className: styles.actionButton,
    style: {
      display: menuOpen ? "inline-block" : void 0,
      marginLeft: getSubrowIndentation()
    },
    "aria-label": ariaLabel,
    "data-header-row": dataHeaderRow,
    "data-column": dataColumn,
    "data-trigger": "useraction",
    children: /* @__PURE__ */ jsx(
      DotMenuIcon,
      {
        style: {
          width: DATA_TABLE_DEFAULT_ICON_SIZE,
          height: DATA_TABLE_DEFAULT_ICON_SIZE
        }
      }
    )
  };
  if (!menuOpen) {
    if (hovered === false) {
      return;
    }
    return /* @__PURE__ */ jsx(
      "button",
      {
        ...triggerProps,
        onClick: (event) => {
          event.preventDefault();
          event.stopPropagation();
          setMenuOpen(true);
          if (type === "cell") {
            table.setOpenMenuRowId(headerGroupOrRowId);
          }
        }
      }
    );
  }
  const alignment = column?.columnDef.alignment?.includes("right") ? "start" : "end";
  return /* @__PURE__ */ jsxs(
    Menu,
    {
      open: menuOpen,
      onOpenChange: (open) => {
        if (!open && onClose) {
          onClose();
        }
        if (!open && type === "cell") {
          table.setOpenMenuRowId(null);
        }
        setMenuOpen(open);
      },
      modal: false,
      children: [
        /* @__PURE__ */ jsx(Menu.Trigger, { children: /* @__PURE__ */ jsx(
          "button",
          {
            ...triggerProps,
            onClick: (evt) => {
              evt.stopPropagation();
            }
          }
        ) }),
        menuOpen === true ? /* @__PURE__ */ jsx(DataTableColumnContextProvider, { value: column, children: /* @__PURE__ */ jsx(Menu.Content, { alignment, children: Children.map(actions, (child) => child) }) }) : null
      ]
    }
  );
}
export {
  InlineUserActionsMenu
};
