// @dynatrace/strato-components/esm/tables/DataTable/features/UserActions/RowActions.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import {
  Children,
  Fragment as Fragment2,
  isValidElement,
  memo,
  useMemo,
  useState
} from "react";
import { useIntl } from "react-intl";
import { isElement } from "react-is";
import { DotMenuIcon } from "@dynatrace/strato-icons";
import * as styles from "./RowActions.sty.js";
import { Button } from "../../../../buttons/button/Button.js";
import { Menu } from "../../../../navigation/menu/Menu.js";
import { isTrigger as isMenuTrigger } from "../../../../navigation/menu/Trigger.js";
import { DataTableDefaultCell } from "../../components/DataTableDefaultCell.js";
import { DATA_TABLE_ROW_ACTIONS_COLUMN_ID } from "../../constants.js";
import { useDataTable } from "../../contexts/DataTableContext.js";
import { dataTableMessages } from "../../messages.js";
function DataTableRowActions(props) {
  return null;
}
function isDataTableRowActions(child) {
  if (isElement(child) && child.type === DataTableRowActions) {
    return true;
  }
  return false;
}
const DataTableInternalRowActions = memo(
  (props) => {
    const table = useDataTable();
    const intl = useIntl();
    const density = table.getVisualVariantOptions().rowDensity === "condensed" ? "condensed" : "default";
    const ariaLabel = intl.formatMessage(dataTableMessages.rowActionsTrigger);
    const [open, setOpen] = useState(false);
    const handleOpenChange = (nextOpen) => {
      setOpen(nextOpen);
      table.setOpenMenuRowId(nextOpen ? props.rowId : null);
    };
    const { menu, freeformSlot } = useMemo(() => {
      const slots = {
        menu: null,
        freeformSlot: []
      };
      const childrenArray = Children.toArray(props.children);
      for (const child of childrenArray) {
        if (isValidElement(child) && child.type === Menu) {
          slots.menu = child;
          continue;
        }
        slots.freeformSlot.push(child);
      }
      return slots;
    }, [props.children]);
    if (!menu) {
      return freeformSlot;
    }
    if (open === false) {
      return /* @__PURE__ */ jsxs(Fragment, { children: [
        freeformSlot,
        /* @__PURE__ */ jsx(
          "span",
          {
            onClick: () => handleOpenChange(true),
            style: { display: "contents" },
            children: getMenuTrigger(menu, density, ariaLabel)
          }
        )
      ] });
    }
    return /* @__PURE__ */ jsxs(Fragment, { children: [
      freeformSlot,
      /* @__PURE__ */ jsxs(
        Menu,
        {
          open,
          onOpenChange: handleOpenChange,
          modal: false,
          ...menu.props,
          children: [
            /* @__PURE__ */ jsx(Menu.Trigger, { children: /* @__PURE__ */ jsx(Button, { variant: "default", size: density, "aria-label": ariaLabel, children: /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(DotMenuIcon, {}) }) }) }),
            menu.props.children
          ]
        }
      )
    ] });
  }
);
function getMenuTrigger(menu, size, ariaLabel) {
  const menuChildren = Children.toArray(menu.props.children);
  const customMenuTrigger = menuChildren.find(isMenuTrigger);
  if (customMenuTrigger) {
    return customMenuTrigger.props.children;
  }
  return getDefaultMenuTrigger(size, ariaLabel);
}
function getDefaultMenuTrigger(size, ariaLabel) {
  return /* @__PURE__ */ jsx(Button, { variant: "default", size, "aria-label": ariaLabel, children: /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(DotMenuIcon, {}) }) });
}
function isFragment(child) {
  return child.type === Fragment2;
}
const ROW_ACTIONS_COLUMN = {
  id: DATA_TABLE_ROW_ACTIONS_COLUMN_ID,
  accessorFn: () => "",
  header: "",
  width: "content",
  enableSorting: false,
  enableUserActions: false,
  enableHiding: false,
  enableResizing: false,
  meta: {
    builtin: true,
    builtinPlacement: "suffix",
    hasCustomCellRenderer: true,
    hasCustomHeaderRenderer: true
  },
  cell: ({ row, table }) => {
    if (table.options.rowActions === void 0) {
      return null;
    }
    const actions = table.options.rowActions(row.original, {
      rowDensity: table.getVisualVariantOptions().rowDensity,
      rowIndex: row.index,
      rowId: row.id
    });
    if (isElement(actions)) {
      return /* @__PURE__ */ jsx(DataTableDefaultCell, { className: styles.rowActions, children: /* @__PURE__ */ jsx(DataTableInternalRowActions, { rowId: row.id, children: isFragment(actions) ? actions.props.children : actions }) });
    }
  }
};
export {
  DataTableRowActions,
  ROW_ACTIONS_COLUMN,
  isDataTableRowActions
};
