// @dynatrace/strato-components/esm/tables/DataTable/features/UserActions/UserActionsMenu.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { useContext } from "react";
import { useIntl } from "react-intl";
import {
  ArrowLeftIcon,
  ArrowRightIcon,
  CopyIcon,
  PinIcon,
  PinOffIcon,
  ViewOffIcon,
  WrapTextIcon
} from "@dynatrace/strato-icons";
import { useCopyToClipboard } from "../../../../core/utils/useCopyToClipboard.js";
import { Group as MenuGroup } from "../../../../navigation/menu/Group.js";
import { Intent as MenuIntent } from "../../../../navigation/menu/Intent.js";
import { Item as MenuItem } from "../../../../navigation/menu/Item.js";
import { Label as MenuLabel } from "../../../../navigation/menu/Label.js";
import { Link as MenuLink } from "../../../../navigation/menu/Link.js";
import { Prefix as MenuPrefix } from "../../../../navigation/menu/Prefix.js";
import { Sub as MenuSub } from "../../../../navigation/menu/Sub.js";
import { SubContent as MenuSubContent } from "../../../../navigation/menu/SubContent.js";
import { SubTrigger as MenuSubTrigger } from "../../../../navigation/menu/SubTrigger.js";
import { Suffix as MenuSuffix } from "../../../../navigation/menu/Suffix.js";
import {
  useDataTable,
  useDataTableColumn,
  useDataTableHeader,
  DataTableHeaderContext
} from "../../contexts/DataTableContext.js";
import { dataTableMessages } from "../../messages.js";
const CopyItem = ({ value, onSelect }) => {
  const intl = useIntl();
  const [copyToClipboard] = useCopyToClipboard(value);
  const header = useContext(DataTableHeaderContext);
  const isColumnAction = header !== null;
  const dataDtName = isColumnAction ? "ColumnActionCopyItem" : "CellActionCopyItem";
  return /* @__PURE__ */ jsxs(
    DataTableActionsMenu.Item,
    {
      onSelect: () => {
        copyToClipboard();
        onSelect?.();
      },
      "data-dt-name": dataDtName,
      children: [
        /* @__PURE__ */ jsx(DataTableActionsMenu.Prefix, { children: /* @__PURE__ */ jsx(CopyIcon, {}) }),
        intl.formatMessage(dataTableMessages.copyValueCellAction)
      ]
    }
  );
};
const LineWrapAction = ({
  onSelect
}) => {
  const intl = useIntl();
  const column = useDataTableColumn();
  const isLineWrapped = column?.getIsLineWrapped();
  const lineWrapDtProperties = `state:${isLineWrapped ? "unwrap" : "wrap"}`;
  return /* @__PURE__ */ jsxs(
    DataTableActionsMenu.Item,
    {
      onSelect: () => {
        column.toggleLineWrap();
        onSelect?.();
      },
      "data-dt-name": "ColumnActionLineWrap",
      "data-dt-properties": lineWrapDtProperties,
      children: [
        /* @__PURE__ */ jsx(DataTableActionsMenu.Prefix, { children: /* @__PURE__ */ jsx(WrapTextIcon, {}) }),
        isLineWrapped ? intl.formatMessage(dataTableMessages.disableLineWrapLabel) : intl.formatMessage(dataTableMessages.enableLineWrapLabel)
      ]
    }
  );
};
const HideColumnAction = ({
  onSelect
}) => {
  const intl = useIntl();
  const table = useDataTable();
  const currentVisibility = table.getState().columnVisibility;
  const column = useDataTableColumn();
  const childColumns = column.columns;
  const handleSelect = () => {
    if (childColumns.length === 0) {
      column.toggleVisibility();
    } else {
      const updatedChildrenVisibility = childColumns.reduce(
        (updatedVisibility, childColumn) => {
          return {
            ...updatedVisibility,
            [childColumn.id]: false
          };
        },
        {}
      );
      table.setColumnVisibility({
        ...currentVisibility,
        ...updatedChildrenVisibility
      });
    }
  };
  return /* @__PURE__ */ jsxs(
    DataTableActionsMenu.Item,
    {
      onSelect: () => {
        handleSelect();
        onSelect?.();
      },
      disabled: !column.getCanHide(),
      "data-dt-name": "ColumnActionHide",
      children: [
        /* @__PURE__ */ jsx(DataTableActionsMenu.Prefix, { children: /* @__PURE__ */ jsx(ViewOffIcon, {}) }),
        intl.formatMessage(dataTableMessages.hideColumnLabel)
      ]
    }
  );
};
const ColumnOrderAction = ({
  onSelect
}) => {
  const intl = useIntl();
  const column = useDataTableColumn();
  const header = useDataTableHeader();
  const isPinnedColumn = !!column.getIsPinned();
  const table = useDataTable();
  if (!table.options.enableColumnOrdering) {
    return null;
  }
  const orderActions = () => /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsxs(
      DataTableActionsMenu.Item,
      {
        onSelect: () => {
          column.moveColumnLeft();
          onSelect?.();
        },
        disabled: !column.getCanMoveColumnLeft(),
        "data-dt-name": "ColumnActionMoveLeft",
        children: [
          /* @__PURE__ */ jsx(DataTableActionsMenu.Prefix, { children: /* @__PURE__ */ jsx(ArrowLeftIcon, {}) }),
          intl.formatMessage(dataTableMessages.moveColumnLeftLabel)
        ]
      }
    ),
    /* @__PURE__ */ jsxs(
      DataTableActionsMenu.Item,
      {
        onSelect: () => {
          column.moveColumnRight();
          onSelect?.();
        },
        disabled: !column.getCanMoveColumnRight(),
        "data-dt-name": "ColumnActionMoveRight",
        children: [
          /* @__PURE__ */ jsx(DataTableActionsMenu.Prefix, { children: /* @__PURE__ */ jsx(ArrowRightIcon, {}) }),
          intl.formatMessage(dataTableMessages.moveColumnRightLabel)
        ]
      }
    )
  ] });
  if (isPinnedColumn) {
    if (header.id.includes("center")) {
      return orderActions();
    }
    return null;
  }
  return orderActions();
};
const ColumnPinningAction = ({
  onSelect
}) => {
  const intl = useIntl();
  const table = useDataTable();
  const header = useDataTableHeader();
  if (!table.options.enableColumnPinning) {
    return null;
  }
  const pinnedState = header.getIsPinned();
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    pinnedState !== "left" ? /* @__PURE__ */ jsxs(
      DataTableActionsMenu.Item,
      {
        onSelect: () => {
          header.pin("left");
          onSelect?.();
        },
        "data-dt-name": "ColumnActionPinLeft",
        children: [
          /* @__PURE__ */ jsx(DataTableActionsMenu.Prefix, { children: /* @__PURE__ */ jsx(PinIcon, {}) }),
          intl.formatMessage(dataTableMessages.pinColumnLeftLabel)
        ]
      }
    ) : null,
    pinnedState !== "right" ? /* @__PURE__ */ jsxs(
      DataTableActionsMenu.Item,
      {
        onSelect: () => {
          header.pin("right");
          onSelect?.();
        },
        "data-dt-name": "ColumnActionPinRight",
        children: [
          /* @__PURE__ */ jsx(DataTableActionsMenu.Prefix, { children: /* @__PURE__ */ jsx(PinIcon, {}) }),
          intl.formatMessage(dataTableMessages.pinColumnRightLabel)
        ]
      }
    ) : null,
    pinnedState !== false ? /* @__PURE__ */ jsxs(
      DataTableActionsMenu.Item,
      {
        onSelect: () => {
          header.pin(false);
          onSelect?.();
        },
        "data-dt-name": "ColumnActionUnpin",
        children: [
          /* @__PURE__ */ jsx(DataTableActionsMenu.Prefix, { children: /* @__PURE__ */ jsx(PinOffIcon, {}) }),
          intl.formatMessage(dataTableMessages.unpinColumnLabel)
        ]
      }
    ) : null
  ] });
};
const ColumnFontStyleAction = ({
  onSelect
}) => {
  const intl = useIntl();
  const column = useDataTableColumn();
  return /* @__PURE__ */ jsxs(
    DataTableActionsMenu.Item,
    {
      onSelect: () => {
        column.toggleFontStyle();
        onSelect?.();
      },
      children: [
        /* @__PURE__ */ jsx(DataTableActionsMenu.Prefix, { children: /* @__PURE__ */ jsx(WrapTextIcon, {}) }),
        column.getFontStyle() === "text" ? intl.formatMessage(dataTableMessages.fontStyleCodeLabel) : intl.formatMessage(dataTableMessages.fontStyleTextLabel)
      ]
    }
  );
};
const DataTableActionsMenu = Object.assign(
  function DataTableActionsMenuInternal(props) {
    return /* @__PURE__ */ jsx(Fragment, { children: props.children });
  },
  {
    Intent: MenuIntent,
    Item: MenuItem,
    // DEPRECATION: [APPDEV-15604] BREAKING-CHANGE: Remove TableActionsMenu.ItemIcon.
    /**
     * The DataTableActionsMenu.ItemIcon is deprecated, use DataTableActionsMenu.Prefix instead.
     * @deprecated Use `TableActionsMenu.Prefix` instead. This slot will be removed in the future.
     */
    ItemIcon: MenuPrefix,
    Prefix: MenuPrefix,
    Suffix: MenuSuffix,
    Group: MenuGroup,
    Label: MenuLabel,
    CopyItem,
    SubMenu: MenuSub,
    SubContent: MenuSubContent,
    SubTrigger: MenuSubTrigger,
    Link: MenuLink,
    LineWrap: LineWrapAction,
    HideColumn: HideColumnAction,
    ColumnOrder: ColumnOrderAction,
    ColumnPinning: ColumnPinningAction,
    ColumnFontStyle: ColumnFontStyleAction
  }
);
export {
  ColumnOrderAction,
  DataTableActionsMenu,
  HideColumnAction
};
