// @dynatrace/strato-components/esm/tables/DataTable/features/UserActions/UserActions.js
import { jsx } from "react/jsx-runtime";
import { isElement } from "react-is";
import {
  filterActionsForChartColumn,
  getColumnActionsElements
} from "./user-actions-utils.js";
import { getMemoOptions } from "../../hooks/useTable/utils.js";
import { memo } from "../../tanstack-table-core-override/src/index.js";
import { SortingColumnActions } from "../Sorting/SortingColumnActions.js";
const UserActions = {
  getInitialState(state) {
    return {
      ...state,
      floatingMenuInfo: {
        columnId: null,
        rowId: null,
        x: 0,
        y: 0
      },
      openMenuRowId: null
    };
  },
  createTable(table) {
    table._setFloatingMenuInfo = (headerId, columnId, rowId, x, y) => {
      table.setState((state) => {
        const newState = { ...state };
        newState.floatingMenuInfo = {
          type: rowId === null ? "headerAction" : "cellAction",
          headerId,
          rowId,
          columnId,
          x,
          y
        };
        newState.openMenuRowId = rowId;
        return newState;
      });
    };
    table.getUserActionsForColumn = (columnId) => {
      const column = table.getColumn(columnId);
      if (column?.getCanColumnActions()) {
        return column.getColumnActions();
      }
      return null;
    };
    table.getUserActionsForCell = (columnId, rowId) => {
      const row = table.getRow(rowId);
      const cell = row.getVisibleCells().find((cell2) => cell2.column.id === columnId);
      if (cell?.column.getCanCellActions()) {
        return cell.getCellActions();
      }
      return null;
    };
    table.getUserActionFloatingMenuContextMenuHandler = () => {
      return (event) => {
        event.preventDefault();
        event.stopPropagation();
        if (event.target instanceof HTMLElement === false && event.target instanceof SVGElement === false) {
          return;
        }
        const cellElement = event.target.closest('[role="cell"]');
        if (cellElement) {
          const { columnid } = cellElement.dataset;
          const rowid = cellElement.closest('[role="row"]')?.getAttribute("data-rowid");
          if (!columnid || !rowid) {
            return;
          }
          const row = table.getRow(rowid);
          const cell = row._getAllCellsByColumnId()[columnid];
          if (cell?.column.getCanCellActions()) {
            table._setFloatingMenuInfo(
              null,
              // header is not needed for floating menu on body cell
              columnid,
              rowid,
              event.pageX,
              event.pageY
            );
          }
          return;
        }
        const headerElement = event.target.closest(
          '[role="columnheader"]'
        );
        if (headerElement) {
          const { columnid, headerid } = headerElement.dataset;
          if (!columnid) {
            return;
          }
          const column = table.getColumn(columnid);
          if (column?.getCanColumnActions()) {
            table._setFloatingMenuInfo(
              headerid ?? null,
              columnid,
              null,
              event.pageX,
              event.pageY
            );
          }
        }
      };
    };
    table.setOpenMenuRowId = (rowId) => {
      table.setState((state) => ({
        ...state,
        openMenuRowId: rowId
      }));
    };
    table.getOpenMenuRowId = () => {
      return table.getState().openMenuRowId;
    };
  },
  createColumn(column, table) {
    column.getColumnActions = memo(
      () => [
        column.id,
        table.options.columnActions,
        table.options.enableSorting
      ],
      (columnId, columnActionsOption, enableSortingOption) => {
        const columnActions = getColumnActionsElements(
          columnId,
          columnActionsOption ?? [],
          table.options.consumerColumnDefMap
        );
        const isChartColumn = typeof column.columnDef.enableUserActions === "object" && column.columnDef.enableUserActions.type === "chart";
        const filteredActions = isChartColumn ? filterActionsForChartColumn(columnActions) : columnActions;
        if (enableSortingOption && column.getCanSort()) {
          filteredActions.push(
            /* @__PURE__ */ jsx(
              SortingColumnActions,
              {
                hasOtherColumnActions: filteredActions.length > 0
              },
              "strato-table-sorting-column-actions"
            )
          );
        }
        return filteredActions;
      },
      getMemoOptions(table.options, "debugColumns", "column.getColumnActions")
    );
    column.getCanColumnActions = () => {
      if (column.columnDef.enableUserActions === false) {
        return false;
      }
      const columnActionsEnabled = table.options.enableColumnActions;
      const sortingEnabled = table.options.enableSorting && column.getCanSort();
      if (!columnActionsEnabled && !sortingEnabled) {
        return false;
      }
      const definedColumnActions = column.getColumnActions();
      if (!definedColumnActions) {
        return false;
      }
      return (definedColumnActions.length ?? 0) > 0;
    };
    column.getCellActionsContent = memo(
      () => [column.id, table.options.cellActions],
      (columnId, cellActionsOptions = []) => {
        return cellActionsOptions.reduce((actions, cellAction) => {
          if (cellAction.column === void 0 || // global action for all columns
          cellAction.column === columnId || // column-specific action
          typeof cellAction.column === "function" && cellAction.column(columnId)) {
            actions.push(cellAction.action);
          }
          return actions;
        }, []);
      },
      getMemoOptions(
        table.options,
        "debugCells",
        "column.getCellActionsElements"
      )
    );
    column.getCanCellActions = () => {
      if (column.columnDef.enableUserActions === false) {
        return false;
      }
      if (!table.options.enableCellActions) {
        return false;
      }
      const definedCellActions = column.getCellActionsContent();
      if (!definedCellActions) {
        return false;
      }
      return (definedCellActions.length ?? 0) > 0;
    };
  },
  createCell(cell, column, row, table) {
    cell.getCellActions = memo(
      () => [column.id, table.options.cellActions],
      () => {
        const cellActions = column.getCellActionsContent().map((cellAction) => {
          if (isElement(cellAction)) {
            return cellAction;
          }
          if (typeof cellAction === "function") {
            const columnDef = table.options.consumerColumnDefMap?.get(
              column.id
            );
            if (columnDef !== void 0) {
              return cellAction({
                columnDef,
                cellValue: cell.getValue(),
                row: row.original,
                rowId: row.id
              });
            }
          }
          return null;
        });
        return cellActions;
      },
      getMemoOptions(table.options, "debugCells", "cell.getCellActions")
    );
  }
};
export {
  UserActions
};
