// @dynatrace/strato-components/esm/tables/DataTable/features/UserActions/useUserActions.js
import {
  useMemo,
  Children
} from "react";
import {
  isDataTableCellActions
} from "./CellActions.js";
import {
  isDataTableColumnActions
} from "./ColumnActions.js";
import { isDataTableRowActions } from "./RowActions.js";
function useUserActions(props, options, consumerColumnDefMap) {
  const configuredActions = useMemo(() => {
    const actions = {
      cellActions: [],
      columnActions: [],
      rowActions: void 0
    };
    Children.forEach(props.children, (child) => {
      if (isDataTableCellActions(child)) {
        actions.cellActions.push({
          column: child.props.column,
          action: child.props.children
        });
      }
      if (isDataTableColumnActions(child)) {
        actions.columnActions.push({
          column: child.props.column,
          action: child.props.children
        });
      }
      if (isDataTableRowActions(child)) {
        actions.rowActions = child.props.children;
      }
    });
    return actions;
  }, [props.children]);
  options.consumerColumnDefMap = consumerColumnDefMap;
  options.enableColumnActions = configuredActions.columnActions.length > 0;
  options.columnActions = configuredActions.columnActions;
  options.enableCellActions = configuredActions.cellActions.length > 0;
  options.cellActions = configuredActions.cellActions;
  options.enableRowActions = configuredActions.rowActions !== void 0;
  options.rowActions = configuredActions.rowActions;
}
export {
  useUserActions
};
