// @dynatrace/strato-components/esm/content/ai-response/AiResponse.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import { useReducedMotion } from "framer-motion";
import { forwardRef, memo } from "react";
import { Streamdown } from "streamdown";
import * as styles from "./AiResponse.sty.js";
import {
  ANIMATION_DURATION_MS,
  useTextAnimation
} from "./hooks/useTextAnimation.js";
import { streamdownComponents } from "./utils/component-mappings.js";
import { useAriaLabelingProps } from "../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useSafeForwardProps } from "../../core/hooks/useSafeForwardProps.js";
const COMPONENT_NAME = "AiResponse";
const animationOptions = {
  animation: "aiResponseAnimation",
  duration: ANIMATION_DURATION_MS,
  easing: "cubic-bezier(0, 0, 0.2, 1)",
  sep: "word"
};
const InternalAiResponse = forwardRef(
  (props, forwardedRef) => {
    const {
      children,
      onAnimationStateChange,
      responseState = "streaming",
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
    const ariaLabelingProps = useAriaLabelingProps(restProps);
    const forwardProps = useSafeForwardProps(restProps, ariaLabelingProps);
    const shouldReduceMotion = useReducedMotion() ?? false;
    const { displayedText, isTextAnimating } = useTextAnimation(
      children,
      shouldReduceMotion,
      responseState,
      onAnimationStateChange
    );
    if (displayedText.length === 0) {
      return null;
    }
    return /* @__PURE__ */ jsx(
      "div",
      {
        ref: forwardedRef,
        "aria-live": "polite",
        "aria-atomic": "true",
        style: consumerStyle,
        className: clsx(styles.wrapper, consumerClassName),
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        ...ariaLabelingProps,
        ...behavioralTrackingProps,
        ...forwardProps,
        "data-dt-component": COMPONENT_NAME,
        children: /* @__PURE__ */ jsx(
          Streamdown,
          {
            animated: shouldReduceMotion ? false : animationOptions,
            isAnimating: responseState === "streaming" || isTextAnimating,
            components: streamdownComponents,
            className: styles.streamdown,
            children: displayedText
          }
        )
      }
    );
  }
);
InternalAiResponse.displayName = COMPONENT_NAME;
const AiResponse = memo(InternalAiResponse);
export {
  AiResponse
};
