// @dynatrace/strato-components/esm/charts/meter-bar/multi-meter-bar/utils/build-multi-meter-bar-chart-config.js
import { merge } from "lodash-es";
import { DEFAULT_FORMATTER } from "../../shared/constants.js";
import {
  defaultToolbarInternalConfig
} from "../../shared/types/meter-bar-toolbar.js";
import { getNumericValue } from "../../shared/utils/get-numeric-value.js";
import { defaultConfig } from "../default.js";
import {
  multiMeterBarChartTooltipVariantOptions
} from "../types/multi-meter-bar-chart.js";
const getToolbarConfig = (toolbarSlotsConfig, toolbarConsumerConfig) => {
  const configuredToolbar = toolbarSlotsConfig ?? toolbarConsumerConfig;
  return { ...defaultToolbarInternalConfig, ...configuredToolbar };
};
const getMaxValue = (totalValue, maxProperty, consumerConfigMax) => {
  if (typeof maxProperty === "number") {
    return maxProperty;
  }
  if (maxProperty === "auto") {
    return totalValue;
  }
  if (!maxProperty && consumerConfigMax) {
    return consumerConfigMax;
  }
  return totalValue;
};
const isValidTooltipVariant = (tooltipVariant) => multiMeterBarChartTooltipVariantOptions.includes(tooltipVariant);
const getTooltipConfig = (slots, consumerConfig) => {
  const consumerTooltip = consumerConfig.tooltip ? {
    ...consumerConfig.tooltip.enabled !== void 0 && {
      hidden: !consumerConfig.tooltip.enabled
    },
    variant: consumerConfig.tooltip.variant
  } : void 0;
  const merged = merge(
    {},
    defaultConfig.tooltip,
    consumerTooltip,
    slots.tooltip
  );
  const variant = isValidTooltipVariant(merged.variant) ? merged.variant : defaultConfig.tooltip.variant;
  return { ...merged, variant };
};
const buildMultiMeterBarChartConfig = (props, consumerConfig, slots) => {
  const size = props.size ?? consumerConfig.size ?? defaultConfig.size;
  const colorPalette = props.colorPalette ?? consumerConfig.colorPalette ?? defaultConfig.colorPalette;
  const min = props.min ?? consumerConfig.min ?? defaultConfig.min;
  const tooltip = getTooltipConfig(slots, consumerConfig);
  const formatter = props.formatter ?? consumerConfig.formatter ?? DEFAULT_FORMATTER;
  const unit = props.unit ?? consumerConfig.unit;
  const legend = slots.legend ?? consumerConfig.legend ?? defaultConfig.legend;
  const totalValue = slots.segments.reduce((acc, segment) => {
    const value = getNumericValue(segment.value, segment.valueAccessor);
    return acc + value;
  }, 0);
  const max = getMaxValue(totalValue, props.max, consumerConfig.max);
  const colorRules = slots.colorRules ?? consumerConfig.colorRules ?? [];
  const toolbar = getToolbarConfig(slots.toolbar, consumerConfig.toolbar);
  const loading = props.loading ?? consumerConfig.loading ?? defaultConfig.loading;
  return {
    segments: slots.segments,
    colorPalette,
    size,
    min,
    max,
    loading,
    tooltip,
    labelSlot: slots.label,
    valueSlot: slots.value,
    minSlot: slots.min,
    maxSlot: slots.max,
    legend,
    formatter,
    unit,
    totalValue,
    colorRules,
    toolbar
  };
};
export {
  buildMultiMeterBarChartConfig
};
