// @dynatrace/strato-components/esm/charts/categorical-bar/displays/BarChartDisplay.js
import { jsx } from "react/jsx-runtime";
import { getBarStyle } from "./bar-style.js";
import { BarCanvas } from "../../core/components/canvas/components/BarCanvas.js";
import { useGridPlotSize } from "../../core/components/chart-grid/hooks/useGridPlotSize.js";
import { safeGet } from "../../core/utils/safe-get.js";
import { useCategoricalBarChartBaseConfig } from "../hooks/useCategoricalBarChartBaseConfig.js";
import { useCategoricalBarFilteredData } from "../hooks/useCategoricalBarFilteredData.js";
import { useDisplayConfigContext } from "../hooks/useDisplayConfigContext.js";
import { getBarCoordinates } from "../utils/categorical-layout.js";
import { isVertical } from "../utils/layout.js";
const BarChartDisplay = () => {
  const { highlightedDimension, categoryScale, dimensionScale, valueScale } = useDisplayConfigContext();
  const { filteredCategories, isSingleDimension, dimensionMap, categoryMap } = useCategoricalBarFilteredData();
  const { layout, groupMode } = useCategoricalBarChartBaseConfig();
  const { width, height } = useGridPlotSize();
  const vertical = isVertical(layout);
  const { categoryCoord, valueCoord, barSize, barThickness } = getBarCoordinates(
    groupMode,
    categoryScale,
    valueScale,
    dimensionScale,
    isSingleDimension,
    dimensionMap,
    categoryMap
  );
  const widthAccessor = vertical ? barThickness : barSize;
  const heightAccessor = vertical ? barSize : barThickness;
  const color = (dimension) => {
    const { name, color: color2 } = safeGet(dimensionMap, dimension);
    return getBarStyle(name, color2, highlightedDimension).color;
  };
  const opacity = (dimension) => {
    const { name, color: color2 } = safeGet(dimensionMap, dimension);
    return getBarStyle(name, color2, highlightedDimension).opacity;
  };
  return filteredCategories.map((category) => {
    const y = vertical ? valueCoord : (d) => categoryCoord(d, category);
    const x = vertical ? (d) => categoryCoord(d, category) : valueCoord;
    return /* @__PURE__ */ jsx(
      BarCanvas,
      {
        data: category.dimensions,
        canvasWidth: width,
        canvasHeight: height,
        x,
        y,
        width: widthAccessor,
        height: heightAccessor,
        isVertical: vertical,
        strokeWidth: 1,
        color,
        opacity
      },
      category.id
    );
  });
};
BarChartDisplay["displayName"] = "BarChartDisplay";
export {
  BarChartDisplay
};
