// @dynatrace/strato-components/esm/charts/xy-chart/processing-strategies/utils/cast-data/cast-data-types.js
import { isPointThreshold } from "../../../../core/types/type-guards.js";
import { getYValue } from "../../../components/thresholds/get-y-value.js";
import {
  XYAxisTypeLinear,
  XYAxisTypeCategorical,
  XYAxisTypeTime,
  XYAxisTypeLog
} from "../../../types/axis.js";
function isValidDate(d) {
  return d instanceof Date && !isNaN(d.valueOf());
}
const castValueToNumber = {
  Number: (value) => value,
  String: (value) => {
    const castValue = parseFloat(value);
    return isNaN(castValue) ? void 0 : castValue;
  },
  Date: (value) => value.getTime()
};
const castValueToString = {
  Number: (value) => value.toString(),
  String: (value) => value,
  Date: (value) => value.toISOString()
};
const castValueToDate = {
  Number: (value) => {
    const castValue = new Date(value);
    return isValidDate(castValue) ? castValue : void 0;
  },
  String: (value) => {
    const castValue = new Date(value);
    return isValidDate(castValue) ? castValue : void 0;
  },
  Date: (value) => isValidDate(value) ? value : void 0
};
const castsMap = {
  [XYAxisTypeLinear]: castValueToNumber,
  [XYAxisTypeCategorical]: castValueToString,
  [XYAxisTypeTime]: castValueToDate,
  [XYAxisTypeLog]: castValueToNumber
};
function castDataType(value, axisType) {
  return castsMap[axisType][value?.constructor.name]?.(
    value
  );
}
function castDynamicThresholdsDataType(data, xAxisType, yAxisType) {
  return data.reduce((acc, d) => {
    const x = castDataType(d.x, xAxisType);
    const yMin = castDataType(getYValue(d, "min"), yAxisType);
    const yMax = castDataType(getYValue(d, "max"), yAxisType);
    const isValidThreshold = x !== void 0 && yMin !== void 0 && yMax !== void 0;
    if (isValidThreshold) {
      if (isPointThreshold(yMin, yMax)) {
        acc.push({ x, y: yMin });
      } else {
        acc.push({ x, yMin, yMax });
      }
    }
    return acc;
  }, []);
}
function castStaticThresholdsDataType(data, yAxisType) {
  const yMin = castDataType(getYValue(data, "min"), yAxisType);
  const yMax = castDataType(getYValue(data, "max"), yAxisType);
  const isValidThreshold = yMin !== void 0 && yMax !== void 0;
  if (isValidThreshold) {
    if (isPointThreshold(yMin, yMax)) {
      return { y: yMin };
    } else {
      return { yMin, yMax };
    }
  }
  return void 0;
}
export {
  castDataType,
  castDynamicThresholdsDataType,
  castStaticThresholdsDataType
};
