// @dynatrace/strato-components/esm/charts/meter-bar/shared/utils/slot-handlers.js
import { isUndefined } from "lodash-es";
import { isValidColorRule } from "../../../core/slots/ColorRule.js";
import { extractBehaviorTrackingProps } from "../../../core/utils/extract-behavior-tracking-props.js";
function handleLabelSlot(child, slots) {
  slots.label = child.props.children;
}
function handleValueSlot(child, slots) {
  slots.value = {
    value: child.props.children,
    auto: isUndefined(child.props.children)
  };
}
function handleMinSlot(child, slots) {
  slots.min = {
    value: child.props.children,
    auto: isUndefined(child.props.children)
  };
}
function handleMaxSlot(child, slots) {
  slots.max = {
    value: child.props.children,
    auto: isUndefined(child.props.children)
  };
}
function handleColorRuleSlot(child, slots) {
  const colorRule = child.props;
  if (isValidColorRule(colorRule)) {
    slots.colorRules.push(colorRule);
  }
}
function handleIntentSlot(child, intents) {
  return intents ? [...intents, child.props] : [child.props];
}
function handleToolbarSlot(child, slots) {
  if (slots.toolbar && "hidden" in child.props) {
    slots.toolbar.hidden = child.props.hidden;
  }
  if (slots.toolbar) {
    Object.assign(slots.toolbar, extractBehaviorTrackingProps(child.props));
  }
}
export {
  handleColorRuleSlot,
  handleIntentSlot,
  handleLabelSlot,
  handleMaxSlot,
  handleMinSlot,
  handleToolbarSlot,
  handleValueSlot
};
