// @dynatrace/strato-components/esm/tables/DataTable/features/ColumnOrder/column-order-utils.js
import { getIsColumnPinned } from "../ColumnPinning/column-pinning-utils.js";
function canMoveColumnLeft(column, columnsToCompare) {
  let columnId = column.id;
  if (column.columns.length > 0) {
    columnId = columnsToCompare.find(
      (leafColumn) => leafColumn.id === column.id || leafColumn.parent?.id === column.id
    )?.id;
  }
  return columnsToCompare.at(0)?.id !== columnId;
}
function canMoveColumnRight(column, columnsToCompare) {
  let columnId = column.id;
  if (column.columns.length > 0) {
    columnId = columnsToCompare.findLast(
      (leafColumn) => leafColumn.id === column.id || leafColumn.parent?.id === column.id
    )?.id;
  }
  return columnsToCompare?.at(-1)?.id !== columnId;
}
function getFirstColumnIndex(column, columnOrder) {
  if (column.columns.length) {
    return columnOrder.findIndex(
      (columnId) => column.columns.some((column2) => column2.id === columnId)
    );
  }
  return columnOrder.indexOf(column.id);
}
function getLastColumnIndex(column, columnOrder) {
  if (column.columns.length) {
    return columnOrder.findLastIndex(
      (columnId) => column.columns.some((column2) => column2.id === columnId)
    );
  }
  return columnOrder.indexOf(column.id);
}
function findColumnById(columnId, columns) {
  return columns.find((c) => c.id === columnId);
}
function getColumnsToMoveCount(column) {
  return (
    // for nested columns, move all child columns together
    column.columns.length || // for standalone columns, move only itself
    1
  );
}
function isColumnFullyPinnedOrHidden(column, columnsPinned) {
  if (!column) {
    return false;
  }
  if (!column.parent?.columns.length) {
    return getIsColumnPinned(column.id, columnsPinned) || !column.getIsVisible();
  }
  return column?.parent.columns.every(
    (col) => getIsColumnPinned(col.id, columnsPinned) || !col.getIsVisible()
  );
}
const getNewColumnOrder = (currentColumnOrder, startIndex, targetIndex, moveCount = 1) => {
  const newColumnOrder = [...currentColumnOrder];
  const columnsToMove = newColumnOrder.splice(startIndex, moveCount);
  newColumnOrder.splice(targetIndex, 0, ...columnsToMove);
  return newColumnOrder;
};
function getColumnsToCompareForMove(table, column) {
  if (column.columns.length > 0 || !column.parent) {
    return table.getCenterVisibleLeafColumnsWithoutBuiltIn();
  }
  return column.parent.getCenterVisibleLeafColumns();
}
function moveNestedColumnLeft(column, leafColumns, columnOrder, columnsPinned) {
  const columnIndex = columnOrder.indexOf(column.id);
  let targetIndex = columnIndex - 1;
  while (targetIndex >= 0) {
    const targetColumnId = columnOrder[targetIndex];
    const targetColumn = leafColumns.find((col) => col.id === targetColumnId);
    if (targetColumn && !getIsColumnPinned(targetColumn.id, columnsPinned) && targetColumn.getIsVisible()) {
      break;
    }
    targetIndex--;
  }
  return getNewColumnOrder(columnOrder, columnIndex, targetIndex);
}
function moveColumnLeft(column, currentColumnOrder, columnsPinned, leafColumns) {
  if (column.parent) {
    const newColumnOrder2 = moveNestedColumnLeft(
      column,
      leafColumns,
      currentColumnOrder,
      columnsPinned
    );
    return newColumnOrder2;
  }
  const columnIndex = getFirstColumnIndex(column, currentColumnOrder);
  let currentTargetIndex = columnIndex - 1;
  while (currentTargetIndex > 0) {
    const currentTargetId = currentColumnOrder[currentTargetIndex];
    const currentTargetColumn = findColumnById(currentTargetId, leafColumns);
    if (currentTargetColumn?.parent) {
      currentTargetIndex -= currentTargetColumn.parent.columns.length - 1;
    }
    if (isColumnFullyPinnedOrHidden(currentTargetColumn, columnsPinned)) {
      currentTargetIndex--;
      continue;
    }
    break;
  }
  const columnsToMove = getColumnsToMoveCount(column);
  const newColumnOrder = getNewColumnOrder(
    currentColumnOrder,
    columnIndex,
    currentTargetIndex,
    columnsToMove
  );
  return newColumnOrder;
}
function moveNestedColumnRight(column, leafColumns, columnOrder, columnsPinned) {
  const columnIndex = columnOrder.indexOf(column.id);
  let targetIndex = columnIndex + 1;
  while (targetIndex < columnOrder.length) {
    const targetColumnId = columnOrder[targetIndex];
    const targetColumn = leafColumns.find((col) => col.id === targetColumnId);
    if (targetColumn && !getIsColumnPinned(targetColumn.id, columnsPinned) && targetColumn.getIsVisible()) {
      break;
    }
    targetIndex++;
  }
  return getNewColumnOrder(columnOrder, columnIndex, targetIndex);
}
function moveColumnRight(column, currentColumnOrder, columnsPinned, leafColumns) {
  if (column.parent) {
    const newColumnOrder2 = moveNestedColumnRight(
      column,
      leafColumns,
      currentColumnOrder,
      columnsPinned
    );
    return newColumnOrder2;
  }
  const columnIndex = getLastColumnIndex(column, currentColumnOrder);
  let currentTargetIndex = columnIndex + 1;
  while (currentTargetIndex < currentColumnOrder.length - 1) {
    const currentTargetId = currentColumnOrder[currentTargetIndex];
    const currentTargetColumn = findColumnById(currentTargetId, leafColumns);
    if (currentTargetColumn?.parent) {
      currentTargetIndex += currentTargetColumn.parent.columns.length - 1;
    }
    if (isColumnFullyPinnedOrHidden(currentTargetColumn, columnsPinned)) {
      currentTargetIndex++;
      continue;
    }
    break;
  }
  const hasChildren = column.columns.length > 0;
  const startIndex = hasChildren ? (
    // We used the last child index as start index,
    // so we need to adjust it to the first child index here.
    columnIndex - (column.columns.length - 1)
  ) : columnIndex;
  const targetIndex = hasChildren ? (
    // We used the last child to calculate the target index,
    // so we need to adjust it here.
    currentTargetIndex - (column.columns.length - 1)
  ) : currentTargetIndex;
  const columnsToMove = getColumnsToMoveCount(column);
  const newColumnOrder = getNewColumnOrder(
    currentColumnOrder,
    startIndex,
    targetIndex,
    columnsToMove
  );
  return newColumnOrder;
}
export {
  canMoveColumnLeft,
  canMoveColumnRight,
  getColumnsToCompareForMove,
  moveColumnLeft,
  moveColumnRight
};
