// @dynatrace/strato-components/esm/tables/DataTable/features/ColumnOrder/ColumnOrder.js
import {
  functionalUpdate,
  memo,
  orderColumns
} from "@tanstack/react-table";
import { isEqual } from "lodash-es";
import { useCallback, useRef } from "react";
import {
  canMoveColumnLeft,
  canMoveColumnRight,
  getColumnsToCompareForMove,
  moveColumnLeft,
  moveColumnRight
} from "./column-order-utils.js";
import {
  extractColumnOrderAndGroups,
  validateColumnOrder
} from "./column-order-validation.js";
import { useControllableState } from "../../../../core/hooks/useControllableState.js";
import { getMemoOptions } from "../../hooks/useTable/utils.js";
const DataTableColumnOrder = {
  createTable(table) {
    table._getOrderColumnsFn = memo(
      () => {
        const tableState = table.getState();
        return [
          tableState.columnOrder,
          tableState.grouping,
          table.options.groupedColumnMode
        ];
      },
      (columnOrder, grouping, groupedColumnMode) => (columns) => {
        const prefixColumns = [];
        const suffixColumns = [];
        let orderedColumns = columns;
        if (columnOrder.length) {
          const columnsMap = /* @__PURE__ */ new Map();
          columns.forEach((column) => {
            const builtinPlacement = column.columnDef.meta?.builtinPlacement;
            if (builtinPlacement === "prefix") {
              prefixColumns.push(column);
            } else if (builtinPlacement === "suffix") {
              suffixColumns.push(column);
            } else {
              columnsMap.set(column.id, column);
            }
          });
          orderedColumns = [
            ...prefixColumns,
            ...columnOrder.flatMap((id) => columnsMap.get(id) ?? []),
            ...suffixColumns
          ];
        }
        return orderColumns(
          orderedColumns,
          grouping,
          groupedColumnMode
        );
      },
      getMemoOptions(table.options, "debugCells", "cell._getOrderColumnsFn")
    );
    table.getLeafColumnsWithoutBuiltin = () => {
      return table.getAllLeafColumns().filter((leafColumn) => !leafColumn.columnDef.meta?.builtin);
    };
    table.setDraggedHeaderColumn = (column) => {
      table.setState((state) => ({
        ...state,
        draggedHeaderColumn: column
      }));
    };
  },
  createHeader(header, table) {
    header.getColumnDragAndDropMode = () => {
      const isDragAndDropDisabled = table.options.enableColumnOrdering !== true || header.column.columnDef.enableUserActions === false || header.getIsPinned() || !header.column.getCanMoveColumnLeft() && !header.column.getCanMoveColumnRight();
      if (isDragAndDropDisabled) {
        return "none";
      }
      if (!header.column.getCanColumnActions() && !header.getCanSortByHeaderClick()) {
        return "only";
      }
      return "enabled";
    };
    header.getIsDragAndDropDisabled = () => {
      const { draggedHeaderColumn } = table.getState();
      const draggedHeaderParentId = draggedHeaderColumn?.parent?.id;
      const draggedHeaderId = draggedHeaderColumn?.id;
      const isPinnedHeader = !!header.getIsPinned();
      if (draggedHeaderId && isPinnedHeader) {
        return true;
      }
      if (draggedHeaderParentId === void 0 || draggedHeaderParentId === header.column.id || draggedHeaderParentId === header.column.parent?.id) {
        return false;
      } else {
        return true;
      }
    };
  },
  createColumn(column, table) {
    column.getCanMoveColumnLeft = () => {
      const columnsToCompare = getColumnsToCompareForMove(table, column);
      return canMoveColumnLeft(column, columnsToCompare);
    };
    column.getCanMoveColumnRight = () => {
      const columnsToCompare = getColumnsToCompareForMove(table, column);
      return canMoveColumnRight(column, columnsToCompare);
    };
    column.moveColumnLeft = () => {
      const leafColumns = table.getLeafColumnsWithoutBuiltin();
      const currentColumnOrder = table.getState().columnOrder;
      const columnsPinned = table.getState().columnPinning;
      const newColumnOrder = moveColumnLeft(
        column,
        currentColumnOrder,
        columnsPinned,
        leafColumns
      );
      table.setColumnOrder(newColumnOrder);
    };
    column.moveColumnRight = () => {
      const leafColumns = table.getLeafColumnsWithoutBuiltin();
      const currentColumnOrder = table.getState().columnOrder;
      const columnsPinned = table.getState().columnPinning;
      const newColumnOrder = moveColumnRight(
        column,
        currentColumnOrder,
        columnsPinned,
        leafColumns
      );
      table.setColumnOrder(newColumnOrder);
    };
  }
};
function useColumnOrder(props, options) {
  options.enableColumnOrdering = false;
  const { columnOrder: columnOrderFromDefinition, columnGroups } = extractColumnOrderAndGroups(props.columns);
  const lastValidColumnOrder = useRef(
    columnOrderFromDefinition
  );
  const lastColumnOrderFromDefinition = useRef(
    columnOrderFromDefinition
  );
  const [columnOrder, setColumnOrder] = useControllableState({
    value: props.columnOrder,
    // if no order was specified, initialize with column definition order
    defaultValue: props.defaultColumnOrder || columnOrderFromDefinition
  });
  const onColumnOrderChangeCallback = useCallback(
    (updater, trigger = "user") => {
      const nextState = functionalUpdate(
        updater,
        options.state?.columnOrder ?? columnOrder
      );
      if (!isEqual(nextState, columnOrder)) {
        setColumnOrder(nextState);
        props.onColumnOrderChange?.(nextState, trigger);
      }
    },
    [columnOrder, options.state?.columnOrder, props, setColumnOrder]
  );
  if (!props.columnOrdering) {
    return;
  }
  if (props.columnOrder === void 0 && !isEqual(lastColumnOrderFromDefinition.current, columnOrderFromDefinition)) {
    onColumnOrderChangeCallback(columnOrderFromDefinition, "internal");
    lastColumnOrderFromDefinition.current = columnOrderFromDefinition;
  } else if (!isEqual(columnOrder, lastValidColumnOrder.current)) {
    const validColumnOrder = validateColumnOrder(
      columnOrder,
      columnOrderFromDefinition,
      columnGroups
    );
    if (validColumnOrder !== null) {
      onColumnOrderChangeCallback(validColumnOrder, "internal");
      lastValidColumnOrder.current = validColumnOrder;
    }
  }
  options.enableColumnOrdering = true;
  options.state = { ...options.state, columnOrder };
  options.onColumnOrderChange = onColumnOrderChangeCallback;
  options.originalColumnOrder = props.defaultColumnOrder ?? lastValidColumnOrder.current;
}
export {
  DataTableColumnOrder,
  useColumnOrder
};
