// @dynatrace/strato-components/esm/tables/DataTable/features/ColumnOrder/ColumnOrderDragAndDropContext.js
import { jsx, jsxs } from "react/jsx-runtime";
import {
  DndContext as DragAndDropContext,
  DragOverlay,
  MouseSensor,
  TouchSensor,
  useSensor,
  useSensors
} from "@dnd-kit/core";
import clsx from "clsx";
import {
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState
} from "react";
import { createPortal } from "react-dom";
import {
  createHeaderPlaceholder,
  createRestrictColumnDragAndDrop,
  customCollisionDetection,
  getCustomScrollSpeed,
  getDraggableAreaBounds,
  getDraggedHeaderElement,
  indicateInsertPosition,
  SCROLL_PROXIMITY_ZONE,
  unpinnedHeaderFilter,
  updateColumnOrderOnDragEnd
} from "./column-order-dnd-utils.js";
import * as styles from "./ColumnOrder.sty.js";
import { getDataTableBorderWidth } from "../../../utils/data-table-border-width.js";
import { useDataTable } from "../../contexts/DataTableContext.js";
const DRAGGED_HEADER_DEFAULT_WIDTH = 100;
const DRAGGED_HEADER_DEFAULT_HEIGHT = 37;
function ColumnOrderDragAndDropContext({
  children
}) {
  const table = useDataTable();
  const totalHeaderHeight = table.getTotalHeaderHeight();
  const visualVariantOptions = table.getVisualVariantOptions();
  const containedVariant = visualVariantOptions.contained;
  const hasSubHeaders = table.getHeaderGroups().length > 1;
  const verticalDividers = visualVariantOptions.verticalDividers || hasSubHeaders;
  const isColumnPinningEnabled = table.options.enableColumnPinning;
  const pinnedColumns = table.getState().columnPinning;
  const virtualizationContainer = table.getRowVirtualizationContainer();
  const [dragOverlayContent, setDragOverlayContent] = useState(null);
  const pinnedDraggableAreaBoundsRef = useRef({ left: void 0, right: void 0 });
  const [draggedHeaderSize, setDraggedHeaderSize] = useState({
    width: DRAGGED_HEADER_DEFAULT_WIDTH,
    height: DRAGGED_HEADER_DEFAULT_HEIGHT
  });
  const sensors = useSensors(
    useSensor(MouseSensor, {}),
    useSensor(TouchSensor, {})
  );
  const restrictColumnDragAndDrop = useMemo(
    () => createRestrictColumnDragAndDrop(
      totalHeaderHeight,
      containedVariant,
      pinnedColumns,
      hasSubHeaders,
      virtualizationContainer
    ),
    [
      totalHeaderHeight,
      virtualizationContainer,
      containedVariant,
      pinnedColumns,
      hasSubHeaders
    ]
  );
  const placeholderElement = useRef(null);
  const scrollFrameRef = useRef(null);
  const pointerXRef = useRef(0);
  const trackPointerXRef = useRef(null);
  const parentHeaderElementRef = useRef(null);
  const stopScrollLoop = useCallback(() => {
    if (scrollFrameRef.current !== null) {
      cancelAnimationFrame(scrollFrameRef.current);
      scrollFrameRef.current = null;
    }
  }, []);
  const startScrollLoop = useCallback(() => {
    if (scrollFrameRef.current !== null) {
      return;
    }
    const loop = () => {
      if (!virtualizationContainer) {
        scrollFrameRef.current = null;
        return;
      }
      const mouseX = pointerXRef.current;
      const containerRect = virtualizationContainer.getBoundingClientRect();
      const rightBound = pinnedDraggableAreaBoundsRef.current.right ?? containerRect.right;
      const leftBound = pinnedDraggableAreaBoundsRef.current.left ?? containerRect.left;
      const rightPenetration = mouseX - (rightBound - SCROLL_PROXIMITY_ZONE);
      const leftPenetration = leftBound + SCROLL_PROXIMITY_ZONE - mouseX;
      const { scrollLeft, scrollWidth, clientWidth } = virtualizationContainer;
      let maxScrollRight = Infinity;
      let maxScrollLeft = Infinity;
      const parentEl = parentHeaderElementRef.current;
      if (parentEl) {
        const parentRect = parentEl.getBoundingClientRect();
        maxScrollRight = Math.max(0, parentRect.right - rightBound);
        maxScrollLeft = Math.max(0, leftBound - parentRect.left);
      }
      if (rightPenetration > 0 && leftPenetration <= 0) {
        const scrollSpeed = getCustomScrollSpeed(rightPenetration);
        const canScrollRight = scrollLeft < scrollWidth - clientWidth;
        if (canScrollRight && maxScrollRight > 0) {
          virtualizationContainer.scrollLeft += Math.min(
            scrollSpeed,
            Math.ceil(maxScrollRight)
          );
        }
      } else if (leftPenetration > 0 && rightPenetration <= 0) {
        const scrollSpeed = getCustomScrollSpeed(leftPenetration);
        const canScrollLeft = scrollLeft > 0;
        if (canScrollLeft && maxScrollLeft > 0) {
          virtualizationContainer.scrollLeft -= Math.min(
            scrollSpeed,
            Math.ceil(maxScrollLeft)
          );
        }
      }
      scrollFrameRef.current = requestAnimationFrame(loop);
    };
    scrollFrameRef.current = requestAnimationFrame(loop);
  }, [virtualizationContainer]);
  const cleanupDragState = useCallback(() => {
    stopScrollLoop();
    parentHeaderElementRef.current = null;
    if (trackPointerXRef.current) {
      window.removeEventListener("pointermove", trackPointerXRef.current);
      trackPointerXRef.current = null;
    }
  }, [stopScrollLoop]);
  const handleDragStart = ({ active }) => {
    const draggedColumn = active.data.current?.column;
    if (isColumnPinningEnabled) {
      pinnedDraggableAreaBoundsRef.current = getDraggableAreaBounds(
        pinnedColumns,
        virtualizationContainer,
        hasSubHeaders ? 1 : 0
      );
      trackPointerXRef.current = (e) => {
        pointerXRef.current = e.clientX;
      };
      window.addEventListener("pointermove", trackPointerXRef.current, {
        passive: true
      });
      if (draggedColumn.parent) {
        const parentId = draggedColumn.parent.id;
        parentHeaderElementRef.current = virtualizationContainer?.querySelector(
          `[role="columnheader"][data-header-row="0"][data-columnid=${CSS.escape(
            parentId
          )}]${unpinnedHeaderFilter}`
        ) ?? null;
      }
      startScrollLoop();
    }
    table.setDraggedHeaderColumn(draggedColumn);
    setDraggedHeaderSize(table.getHeaderSize(active.id));
    setDragOverlayContent(active.data.current?.portalRef);
    const { headerGroupElement, draggedHeaderElement } = getDraggedHeaderElement(String(active.id), virtualizationContainer);
    if (headerGroupElement === null || !(headerGroupElement instanceof HTMLElement)) {
      return;
    }
    if (draggedHeaderElement === null || !(draggedHeaderElement instanceof HTMLElement)) {
      return;
    }
    const { gridArea } = draggedHeaderElement.style;
    const { width, height } = draggedHeaderElement.getBoundingClientRect();
    if (!gridArea) {
      return;
    }
    const verticalDividerOptions = verticalDividers && !draggedColumn.getIsFirstVisibleUnpinnedColumn() ? {
      isParentColumn: draggedHeaderElement.getAttribute("data-header-row") === "0",
      heightOffset: headerGroupElement.offsetHeight - draggedHeaderElement.offsetHeight
    } : void 0;
    placeholderElement.current = createHeaderPlaceholder(
      gridArea,
      width,
      height,
      verticalDividerOptions
    );
    headerGroupElement.appendChild(placeholderElement.current);
  };
  const handleDragOver = (event) => {
    indicateInsertPosition(event, virtualizationContainer);
  };
  const handleDragEnd = ({
    active,
    over
  }) => {
    setDragOverlayContent(null);
    cleanupDragState();
    updateColumnOrderOnDragEnd({
      active,
      over,
      table,
      virtualizationContainer
    });
    requestAnimationFrame(() => {
      placeholderElement.current?.remove();
    });
  };
  useEffect(() => {
    return () => {
      cleanupDragState();
      placeholderElement.current?.remove();
    };
  }, [stopScrollLoop, cleanupDragState]);
  return /* @__PURE__ */ jsxs(
    DragAndDropContext,
    {
      sensors,
      collisionDetection: customCollisionDetection,
      modifiers: [restrictColumnDragAndDrop],
      autoScroll: !isColumnPinningEnabled,
      accessibility: { container: table.getRowVirtualizationContainer() },
      onDragStart: handleDragStart,
      onDragOver: handleDragOver,
      onDragEnd: handleDragEnd,
      // We need to portal the drag and drop overlay to escape CSS transforms on ancestor elements.
      // This ensures transforms on parent containers don't affect the position of the drag overlay.
      children: [
        createPortal(
          /* @__PURE__ */ jsx(
            DragOverlay,
            {
              "data-drag-overlay": true,
              dropAnimation: null,
              className: clsx(styles.columnDndOverlay),
              style: {
                // the observed header size includes a border which we don't want in the overlay
                height: draggedHeaderSize.height - getDataTableBorderWidth(),
                width: draggedHeaderSize.width
              },
              children: /* @__PURE__ */ jsx("div", { ref: dragOverlayContent })
            }
          ),
          document.body
        ),
        children
      ]
    }
  );
}
export {
  ColumnOrderDragAndDropContext
};
