// @dynatrace/strato-components/esm/tables/DataTable/features/ColumnOrder/column-order-dnd-utils.js
import Spacings from "@dynatrace/strato-design-tokens/spacings";
import * as styles from "./ColumnOrder.sty.js";
import { getDataTableBorderWidth } from "../../../utils/data-table-border-width.js";
import {
  DATA_TABLE_DEFAULT_MIN_COLUMN_WIDTH,
  DATA_TABLE_DND_PLACEHOLDER_DIVIDER_CSS_VAR,
  DATA_TABLE_HORIZONTAL_DIVIDER_HEIGHT
} from "../../constants.js";
const unpinnedHeaderFilter = ':not([data-headerid*="left"]):not([data-headerid*="right"])';
const SCROLL_PROXIMITY_ZONE = 10;
const COLUMN_DND_COLLISION_OFFSET = DATA_TABLE_DEFAULT_MIN_COLUMN_WIDTH / 2 - 1;
const TOUCH_DRAG_MOTION_THRESHOLD = 5;
const MIN_SCROLL_SPEED = 3;
const MAX_SCROLL_SPEED = 15;
function isDragMotion(touchMove, touchStart) {
  const touchStartX = touchStart?.touches[0].clientX;
  if (touchStartX === void 0) {
    return false;
  }
  const touchMoveX = touchMove.touches[0].clientX;
  return Math.abs(touchMoveX - touchStartX) > TOUCH_DRAG_MOTION_THRESHOLD;
}
const updateColumnOrderOnDragEnd = ({
  active,
  over,
  table,
  virtualizationContainer
}) => {
  if (over === null || active.data.current === void 0 || over.data.current === void 0 || virtualizationContainer === void 0) {
    return;
  }
  const activeColumn = active.data.current.column;
  const activeChildColumnIds = activeColumn.columns.map((child) => child.id);
  const activeChildColumnsCount = activeChildColumnIds.length;
  const activeParentColumnId = activeColumn.parent?.id;
  const overColumn = over.data.current.column;
  const overParentColumnId = over.data.current?.column?.parent?.id;
  const overChildColumnIds = overColumn.columns.map((child) => child.id);
  const overChildColumnsCount = overChildColumnIds.length;
  if (activeParentColumnId === overParentColumnId) {
    const columnOrder = table.getState().columnOrder;
    const moveCount = Math.max(activeChildColumnsCount, 1);
    const activeColumnIndex = columnOrder.findIndex((columnId) => {
      const id = activeChildColumnsCount === 0 ? activeColumn.id : (
        // get the index of the first child column if the active column is a parent
        activeColumn.getLeafColumns()[0].id
      );
      return columnId === id;
    });
    const overColumnIndex = columnOrder.findIndex((columnId) => {
      const id = overChildColumnsCount === 0 ? overColumn.id : (
        // get the index of the first child column if the active column is a parent
        overColumn.getLeafColumns()[0].id
      );
      return columnId === id;
    });
    const newColumnOrder = [...columnOrder];
    const movedColumns = newColumnOrder.splice(activeColumnIndex, moveCount);
    const toIndex = overColumnIndex > activeColumnIndex ? overColumnIndex - (moveCount - 1) : overColumnIndex;
    newColumnOrder.splice(toIndex, 0, ...movedColumns);
    table.setColumnOrder(newColumnOrder);
  }
  table.setDraggedHeaderColumn(void 0);
  removeInsertPositionIndicator(lastStyledCells, virtualizationContainer, true);
};
const getCustomScrollSpeed = (scrollZoneDepth) => {
  return Math.round(
    MIN_SCROLL_SPEED + Math.min(scrollZoneDepth, SCROLL_PROXIMITY_ZONE) / SCROLL_PROXIMITY_ZONE * (MAX_SCROLL_SPEED - MIN_SCROLL_SPEED)
  );
};
const createHeaderPlaceholder = (gridArea, width, height, verticalDividerOptions) => {
  const placeholder = document.createElement("div");
  placeholder.style.gridArea = gridArea;
  placeholder.style.width = `${width}px`;
  placeholder.style.height = `${height}px`;
  if (verticalDividerOptions) {
    const { isParentColumn, heightOffset } = verticalDividerOptions;
    placeholder.classList.add(styles.verticalDividers);
    placeholder.style.setProperty(
      DATA_TABLE_DND_PLACEHOLDER_DIVIDER_CSS_VAR,
      !isParentColumn && heightOffset !== 0 ? `${heightOffset}px` : Spacings.Size4
    );
  }
  return placeholder;
};
const lastStyledCells = /* @__PURE__ */ new Set();
function indicateInsertPosition(event, virtualizationContainer) {
  if (virtualizationContainer === void 0 || !event.active || !event.over) {
    return;
  }
  const activeColumnIndex = event.active.data.current?.columnIndex;
  const overColumnId = event.over.id;
  const overColumnIndex = event.over.data.current?.columnIndex;
  const overColumnChildren = (event.over.data.current?.column).getVisibleLeafColumns().filter((child) => !child.getIsPinned());
  const overFirstChildId = overColumnChildren.at(0)?.id;
  const overLastChildId = overColumnChildren.at(-1)?.id;
  if (activeColumnIndex < overColumnIndex) {
    const overChildColumnId = overLastChildId !== overColumnId ? overLastChildId : void 0;
    renderInsertPositionIndicator(
      virtualizationContainer,
      overColumnId,
      overChildColumnId,
      "right"
    );
  } else if (activeColumnIndex > overColumnIndex) {
    const overChildColumnId = overFirstChildId !== overColumnId ? overFirstChildId : void 0;
    renderInsertPositionIndicator(
      virtualizationContainer,
      overColumnId,
      overChildColumnId,
      "left"
    );
  } else {
    removeInsertPositionIndicator(lastStyledCells);
  }
}
function renderInsertPositionIndicator(virtualizationContainer, columnId, childColumnId, position) {
  const overCell = getHeaderCell(virtualizationContainer, columnId);
  if (overCell === null) {
    return;
  }
  const overChildColumnCell = getHeaderCell(
    virtualizationContainer,
    childColumnId
  );
  removeInsertPositionIndicator(lastStyledCells);
  const resizeHandles = getResizeHandles(virtualizationContainer);
  for (const resizeHandle of resizeHandles) {
    resizeHandle.style.display = "none";
  }
  addPositionIndicatorStyle(overCell, position);
  lastStyledCells.add(overCell);
  if (overChildColumnCell) {
    addPositionIndicatorStyle(overChildColumnCell, position);
    lastStyledCells.add(overChildColumnCell);
  }
}
function addPositionIndicatorStyle(elem, position) {
  elem.classList.add(
    styles.insertPositionIndicator.classNames.base,
    styles.insertPositionIndicator.classNames.variants.left[position === "left" ? "true" : "false"]
  );
}
function removeInsertPositionIndicator(cells, virtualizationContainer = void 0, displayResizeHandles = false) {
  if (cells.size <= 0) {
    return;
  }
  if (displayResizeHandles && virtualizationContainer) {
    const resizeHandles = getResizeHandles(virtualizationContainer);
    for (const resizeHandle of resizeHandles) {
      resizeHandle.style.removeProperty("display");
    }
  }
  for (const cell of cells) {
    cell.classList.remove(
      styles.insertPositionIndicator.classNames.base,
      styles.insertPositionIndicator.classNames.variants.left.true,
      styles.insertPositionIndicator.classNames.variants.left.false
    );
    lastStyledCells.delete(cell);
  }
}
function getHeaderCell(virtualizationContainer, columnId) {
  if (columnId === void 0) {
    return null;
  }
  return virtualizationContainer.querySelector(
    `[role="columnheader"][data-headerid=${CSS.escape(
      columnId
    )}], [role='columnheader'][data-columnid="${CSS.escape(columnId)}"]`
  );
}
function getResizeHandles(virtualizationContainer) {
  const resizeHandles = virtualizationContainer.querySelectorAll(
    '[data-trigger="resize"]'
  );
  return Array.from(resizeHandles);
}
function isPointWithinRectX(pointX, rect) {
  return rect.left <= pointX && pointX <= rect.right;
}
const closestTargetWithinX = ({
  droppableContainers,
  droppableRects,
  active,
  collisionRect
}) => {
  const initialRect = active.rect.current.initial;
  if (initialRect === null) {
    return [];
  }
  const collisions = [];
  const isDraggingLeft = collisionRect.left < initialRect.left;
  const collisionRectX = isDraggingLeft ? collisionRect.left + COLUMN_DND_COLLISION_OFFSET : collisionRect.right - COLUMN_DND_COLLISION_OFFSET;
  for (const droppableContainer of droppableContainers) {
    const { id } = droppableContainer;
    const rect = droppableRects.get(id);
    if (rect === void 0) {
      continue;
    }
    if (isPointWithinRectX(collisionRectX, rect)) {
      return [
        {
          id,
          data: { droppableContainer, value: 0 }
        }
      ];
    }
    const droppableContainerX = isDraggingLeft ? rect.left : rect.right;
    const distance = Math.abs(collisionRectX - droppableContainerX);
    collisions.push({ id, data: { droppableContainer, value: distance } });
  }
  return collisions.sort((a, b) => a.data.value - b.data.value);
};
const customCollisionDetection = ({
  active,
  droppableContainers,
  ...rest
}) => {
  if (active.data.current === void 0) {
    return [];
  }
  const activeParentColumnId = active.data.current?.column?.parent?.id;
  let validTargets;
  if (activeParentColumnId === void 0) {
    validTargets = droppableContainers.filter((target) => {
      const targetIsChildColumn = target.data.current?.column?.parent !== void 0;
      return !targetIsChildColumn;
    });
  } else {
    validTargets = droppableContainers.filter((target) => {
      const targetParentColumnId = target.data.current?.column?.parent?.id;
      return targetParentColumnId === activeParentColumnId;
    });
  }
  validTargets = validTargets.filter((target) => {
    const id = String(target.id);
    if (id.startsWith("left_") || id.startsWith("right_")) {
      return false;
    }
    const column = target.data.current?.column;
    if (!column) {
      return false;
    }
    if (column.columnDef.meta?.builtin === true) {
      return false;
    }
    const childColumns = column?.columns ?? [];
    if (childColumns.length > 0) {
      return childColumns.some((child) => !child.getIsPinned());
    }
    return !column.getIsPinned();
  });
  return closestTargetWithinX({
    active,
    droppableContainers: validTargets,
    ...rest
  });
};
function getHeaderBounds(columnId, virtualizationContainer, rowIndex = 0, centerOnly = false) {
  const element = virtualizationContainer.querySelector(
    `[role="columnheader"][data-header-row="${rowIndex}"][data-columnId=${CSS.escape(
      columnId
    )}]${centerOnly ? unpinnedHeaderFilter : ""}`
  );
  if (element) {
    const { left, right, height } = element.getBoundingClientRect();
    return { left, right, height };
  }
  return {
    left: void 0,
    right: void 0,
    height: void 0
  };
}
const getTableBounds = (containedVariant, virtualizationContainer, totalHeaderHeight) => {
  const parentRect = virtualizationContainer.getBoundingClientRect();
  return {
    left: containedVariant ? parentRect.left + getDataTableBorderWidth() : parentRect.left,
    right: containedVariant ? parentRect.right - getDataTableBorderWidth() : parentRect.right,
    top: containedVariant ? parentRect.top + getDataTableBorderWidth() : parentRect.top,
    bottom: parentRect.bottom - totalHeaderHeight
  };
};
function getDraggableAreaBounds(pinnedColumns, virtualizationContainer, rowIndex = 0) {
  let left;
  let right;
  const leftPinnedColumns = pinnedColumns.left;
  const rightPinnedColumns = pinnedColumns.right;
  if (virtualizationContainer) {
    if (leftPinnedColumns?.length) {
      const lastLeftPinnedColumn = leftPinnedColumns[leftPinnedColumns.length - 1];
      left = getHeaderBounds(
        lastLeftPinnedColumn,
        virtualizationContainer,
        rowIndex
      ).right;
    }
    if (rightPinnedColumns?.length) {
      right = getHeaderBounds(
        rightPinnedColumns[0],
        virtualizationContainer,
        rowIndex
      ).left;
    }
  }
  return { left, right };
}
function getDraggedHeaderElement(headerId, virtualizationContainer) {
  if (!virtualizationContainer) {
    return { headerGroupElement: null, draggedHeaderElement: null };
  }
  const headerGroupElement = virtualizationContainer.querySelector('[role="rowgroup"]');
  const draggedHeaderElement = headerGroupElement?.querySelector(
    `[data-headerid="${CSS.escape(headerId)}"]`
  );
  return { headerGroupElement, draggedHeaderElement };
}
function restrictToTableBounds(transform, rect, tableBounds) {
  const value = {
    ...transform
  };
  if (typeof tableBounds.left === "number") {
    if (rect.left + transform.x <= tableBounds.left) {
      value.x = tableBounds.left - rect.left;
    }
  }
  if (typeof tableBounds.right === "number") {
    if (rect.right + transform.x >= tableBounds.right) {
      value.x = tableBounds.right - rect.right;
    }
  }
  if (typeof tableBounds.top === "number") {
    if (rect.top + transform.y <= tableBounds.top) {
      value.y = tableBounds.top - rect.top;
    }
  }
  if (typeof tableBounds.bottom === "number") {
    if (rect.bottom + transform.y >= tableBounds.bottom) {
      value.y = tableBounds.bottom - rect.bottom;
    }
  }
  return value;
}
const createRestrictColumnDragAndDrop = (totalHeaderHeight, containedVariant, pinnedColumns, hasSubHeaders, virtualizationContainer) => {
  return ({ transform, active, draggingNodeRect }) => {
    if (!draggingNodeRect || virtualizationContainer === void 0) {
      return transform;
    }
    const activeColumnParentId = active?.data.current?.column?.parent?.id;
    const hasChildColumns = activeColumnParentId === void 0 && active?.data.current?.column?.columns.length > 0;
    const isChildColumn = activeColumnParentId !== void 0;
    const hasPinnedColumns = !!(pinnedColumns.left?.length ?? pinnedColumns.right?.length);
    const headerRowIndex = hasSubHeaders ? 1 : 0;
    const tableBounds = getTableBounds(
      containedVariant,
      virtualizationContainer,
      totalHeaderHeight
    );
    let leftPinnedAreaBound, rightPinnedAreaBound;
    if (hasPinnedColumns) {
      const { left, right } = getDraggableAreaBounds(
        pinnedColumns,
        virtualizationContainer,
        headerRowIndex
      );
      leftPinnedAreaBound = left;
      rightPinnedAreaBound = right;
      tableBounds.left = left ?? tableBounds.left;
      tableBounds.right = right ?? tableBounds.right;
    }
    const updateTableBounds = (left, right, height) => {
      if (height !== void 0 && tableBounds) {
        const horizontalDividerHeight = hasChildColumns ? DATA_TABLE_HORIZONTAL_DIVIDER_HEIGHT : 0;
        tableBounds.bottom = (tableBounds.bottom ?? 0) + height + horizontalDividerHeight;
      }
      if (isChildColumn && left !== void 0 && right !== void 0) {
        tableBounds.left = Math.max(left, leftPinnedAreaBound ?? left);
        tableBounds.right = Math.min(right, rightPinnedAreaBound ?? right);
      }
    };
    if (hasChildColumns || isChildColumn) {
      const activeColumnId = active?.data.current?.column?.id;
      const columnIdForMeasuringHeaderBounds = hasChildColumns ? activeColumnId : activeColumnParentId;
      const { left, right, height } = getHeaderBounds(
        columnIdForMeasuringHeaderBounds,
        virtualizationContainer,
        0,
        isChildColumn && hasPinnedColumns
      );
      updateTableBounds(left, right, height);
    }
    return restrictToTableBounds(transform, draggingNodeRect, tableBounds);
  };
};
export {
  SCROLL_PROXIMITY_ZONE,
  createHeaderPlaceholder,
  createRestrictColumnDragAndDrop,
  customCollisionDetection,
  getCustomScrollSpeed,
  getDraggableAreaBounds,
  getDraggedHeaderElement,
  indicateInsertPosition,
  isDragMotion,
  restrictToTableBounds,
  unpinnedHeaderFilter,
  updateColumnOrderOnDragEnd
};
