// @dynatrace/strato-components/esm/charts/xy-chart-adoption/histogram/HistogramToXY.js
import { jsx, jsxs } from "react/jsx-runtime";
import {
  forwardRef,
  useCallback,
  useImperativeHandle,
  useMemo,
  useRef
} from "react";
import { useHistogramToXYColorRules } from "./hooks/use-histogram-to-xy-color-rules.js";
import {
  ANNOTATIONS_KEYS,
  transformHistogramAnnotationsToXY
} from "./transformer-annotations-to-xy.js";
import { transformHistogramDataToXYChart } from "./transformer-data-to-xy.js";
import {
  mapSelectedSeriesToHistogram,
  mapSelectionDomainToHistogramDomains
} from "./transformer-selection-to-histogram.js";
import { mapHistogramSeriesActionsToXY } from "./transformer-series-actions-to-xy.js";
import { mapHistogramThresholdsToXY } from "./transformer-thresholds-to-xy.js";
import { buildYAxisProps } from "./utils/axis-props-mapping.js";
import { buildHistogramImperativeHandle } from "./utils/build-histogram-imperative-handle.js";
import { buildYAxesConfigFromConfig } from "./utils/build-y-axes-config.js";
import { defineAxisId } from "./utils/define-axis-id.js";
import { deriveXYChartPropsFromConfig } from "./utils/derive-xychart-props-from-config.js";
import { formatCSV } from "./utils/format-csv.js";
import { downloadCSV } from "../../../core/utils/download.js";
import { getChartStatesTemplates } from "../../core/components/chart-states/utils/get-chart-states-template.js";
import { CSS_DOWNLOAD_FILENAME } from "../../core/constants.js";
import { useHistogramChartConfig } from "../../histogram/hooks/useHistogramChartConfig.js";
import { XYChart, XYChartAnnotations } from "../../xy-chart/XYChartRoot.js";
const getAnnotationField = (d, key) => d[key];
const HistogramToXY = forwardRef((props, forwardedRef) => {
  const {
    data: propsData,
    loading,
    children,
    className,
    style,
    width,
    height,
    colorPalette,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    truncationMode,
    valueRepresentation,
    seriesActions,
    initialZoom,
    currentZoom,
    onZoomChange,
    infiniteZoom
  } = props;
  const config = useHistogramChartConfig(children, {
    colorPalette,
    truncationMode,
    valueRepresentation,
    initialZoom,
    currentZoom,
    onZoomChange,
    infiniteZoom
  });
  const { barSeriesProps, units, seriesColors } = transformHistogramDataToXYChart(propsData);
  const allColorRules = useHistogramToXYColorRules({
    data: propsData,
    seriesColors,
    colorRules: config.colorRules
  });
  const { emptyState, errorState } = getChartStatesTemplates(children);
  const downloadDataCallback = useCallback(() => {
    const csvString = formatCSV(propsData);
    downloadCSV(csvString, CSS_DOWNLOAD_FILENAME);
  }, [propsData]);
  const selectActions = config.select?.actions;
  const mappedSeriesActions = useMemo(
    () => seriesActions ? mapHistogramSeriesActionsToXY(seriesActions, propsData) : void 0,
    [seriesActions, propsData]
  );
  const yAxesConfig = buildYAxesConfigFromConfig(config);
  const mappedThresholds = useMemo(
    () => mapHistogramThresholdsToXY(config.thresholds, [...units]),
    [config.thresholds, units]
  );
  const {
    xAxisConfig,
    disableZoom,
    disablePan,
    toolbarEnabled,
    toolbarPlacement,
    toolbarIntents,
    onZoomChange: effectiveOnZoomChange
  } = deriveXYChartPropsFromConfig(config, onZoomChange);
  const xyRef = useRef(null);
  useImperativeHandle(
    forwardedRef,
    () => buildHistogramImperativeHandle(xyRef, config),
    [config]
  );
  const annotations = config.annotations;
  const xAxisId = "x-axis";
  return /* @__PURE__ */ jsxs(
    XYChart,
    {
      data: [],
      ref: xyRef,
      loading,
      className,
      style,
      width,
      height,
      colorPalette: config.colorPalette,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      truncationMode: config.truncationMode,
      __downloadDataCallback: downloadDataCallback,
      children: [
        errorState && /* @__PURE__ */ jsx(XYChart.ErrorState, { children: errorState }),
        emptyState && /* @__PURE__ */ jsx(XYChart.EmptyState, { children: emptyState }),
        selectActions ? /* @__PURE__ */ jsx(
          XYChart.Select,
          {
            actions: (selectedSeries, selectionDomain) => selectActions(
              mapSelectedSeriesToHistogram(selectedSeries),
              mapSelectionDomainToHistogramDomains(selectionDomain)
            )
          }
        ) : null,
        barSeriesProps.map(({ data, unit }, index) => /* @__PURE__ */ jsx(
          XYChart.BarSeries,
          {
            data,
            xAxisId,
            yAxisId: defineAxisId(unit),
            x0Accessor: "x0",
            x1Accessor: "x1",
            y1Accessor: "y1",
            variant: config.valueRepresentation === "relative" && propsData.length === 1 ? "single" : "stack",
            seriesIdAccessor: "seriesId",
            nameAccessor: "name",
            actions: mappedSeriesActions
          },
          index
        )),
        /* @__PURE__ */ jsx(
          XYChart.XAxis,
          {
            id: xAxisId,
            type: xAxisConfig.scale === "log" ? "log" : "numerical",
            position: "bottom",
            disableZoom,
            disablePan,
            initialZoom: config.initialZoom,
            currentZoom: config.currentZoom,
            onZoomChange: effectiveOnZoomChange,
            infiniteZoom: config.infiniteZoom,
            ...xAxisConfig
          }
        ),
        [...units].map((unit, index) => /* @__PURE__ */ jsx(
          XYChart.YAxis,
          {
            ...buildYAxisProps(
              index,
              unit,
              yAxesConfig,
              config.valueRepresentation
            )
          },
          defineAxisId(unit)
        )),
        /* @__PURE__ */ jsx(XYChart.Legend, { ...config.legend }),
        allColorRules.map((rule) => /* @__PURE__ */ jsx(
          XYChart.ColorRule,
          {
            ...rule
          },
          `${rule.comparator}-${rule.matchValue}-${rule.valueAccessor}`
        )),
        mappedThresholds.map((threshold) => /* @__PURE__ */ jsx(
          XYChart.ThresholdIndicator,
          {
            ...threshold
          },
          `${threshold.yAxisId}-${threshold.color}-${threshold.label ?? ""}`
        )),
        /* @__PURE__ */ jsx(XYChart.Toolbar, { hidden: !toolbarEnabled, placement: toolbarPlacement }),
        toolbarIntents?.map((intentProps, index) => /* @__PURE__ */ jsx(XYChart.Intent, { ...intentProps }, index)),
        annotations && /* @__PURE__ */ jsxs(
          XYChart.Annotations,
          {
            data: transformHistogramAnnotationsToXY(annotations.tracks),
            xAxisId,
            showLabels: annotations.showLabels,
            fixedTracks: annotations.visibleTracksLimit,
            annotationsActions: annotations.annotationsActions,
            children: [
              /* @__PURE__ */ jsx(
                XYChartAnnotations.Track,
                {
                  trackIdAccessor: ANNOTATIONS_KEYS.trackId,
                  labelAccessor: ANNOTATIONS_KEYS.trackLabel,
                  colorAccessor: ANNOTATIONS_KEYS.trackColor,
                  hidden: (d) => getAnnotationField(d, ANNOTATIONS_KEYS.trackHidden),
                  symbol: (d) => getAnnotationField(d, ANNOTATIONS_KEYS.trackSymbol),
                  indicatorsDisplay: (d) => getAnnotationField(
                    d,
                    ANNOTATIONS_KEYS.trackIndicators
                  )
                }
              ),
              /* @__PURE__ */ jsx(
                XYChartAnnotations.Marker,
                {
                  startAccessor: ANNOTATIONS_KEYS.markerStart,
                  endAccessor: ANNOTATIONS_KEYS.markerEnd,
                  trackByAccessor: ANNOTATIONS_KEYS.trackId,
                  titleAccessor: ANNOTATIONS_KEYS.markerTitle,
                  descriptionAccessor: ANNOTATIONS_KEYS.markerDescription,
                  colorAccessor: ANNOTATIONS_KEYS.markerColor,
                  hidden: (d) => getAnnotationField(d, ANNOTATIONS_KEYS.markerHidden),
                  symbol: (d) => getAnnotationField(d, ANNOTATIONS_KEYS.markerSymbol),
                  indicatorsDisplay: (d) => getAnnotationField(
                    d,
                    ANNOTATIONS_KEYS.markerIndicators
                  ),
                  priority: (d) => getAnnotationField(d, ANNOTATIONS_KEYS.markerPriority)
                }
              ),
              annotations.tooltip?.template && /* @__PURE__ */ jsx(XYChartAnnotations.Tooltip, { hidden: annotations.tooltip.hidden, children: annotations.tooltip.template })
            ]
          }
        )
      ]
    }
  );
});
HistogramToXY.displayName = "HistogramToXY";
export {
  HistogramToXY
};
