// @dynatrace/strato-components/esm/charts/xy-chart/components/selection/XYChartSelectionAreaTooltip.js
import { jsx } from "react/jsx-runtime";
import { isUndefined } from "lodash-es";
import { useIntl } from "react-intl";
import { ZoomToSelectionIcon } from "@dynatrace/strato-icons";
import { TooltipHeader } from "./TooltipHeader.js";
import { getCustomContextMenuItems } from "../../../core/components/interactions-overlay/utils/get-custom-context-menu-items.js";
import { useOverlay } from "../../../core/components/overlay-tooltip/contexts/OverlayContext.js";
import { OverlayTooltip } from "../../../core/components/overlay-tooltip/OverlayTooltip.js";
import { SelectionTooltipWithAsyncActions } from "../../../core/components/selection-area/SelectionTooltipWithAsyncActions.js";
import { getSelectionTooltipPosition } from "../../../core/components/selection-area/tooltip/get-selection-tooltip-position.js";
import { usePointerAndScrollResponsiveChartBounds } from "../../../core/hooks/usePointerAndScrollResponsiveChartBounds.js";
import { useSelectConfig } from "../../hooks/useSelectConfig.js";
import { useXYChartData } from "../../hooks/useXYChartData.js";
import { XYCHART_MESSAGES } from "../../messages/messages.js";
import { useSelectState } from "../../store/selectors.js";
import { getDatapointsMessage } from "../../utils/get-datapoints-message.js";
import { useZoomToSelection } from "../overlay/hooks/useZoomToSelection.js";
const XYChartSelectionAreaTooltip = ({
  selectedDatapoints,
  totalSeries: seriesCount,
  totalCategories: categoriesCount,
  selectedDatapointsPerSeries
}) => {
  const selectState = useSelectState();
  const { axesMetadata, seriesWithDatapoints } = useXYChartData();
  const selectConfig = useSelectConfig();
  if (isUndefined(selectState)) {
    return null;
  }
  const selectedDatapointsWithData = selectedDatapointsPerSeries.filter(
    (data) => data.datapoints.length
  );
  const selectedSeries = selectedDatapointsWithData.map((data) => {
    return {
      [data.seriesName]: data.datapoints
    };
  });
  const selectionDomain = selectedDatapointsWithData.map((data) => {
    return {
      [data.seriesName]: data.domain
    };
  });
  const datapointsMessage = getDatapointsMessage(
    selectedDatapoints,
    seriesCount,
    seriesWithDatapoints.length,
    categoriesCount
  );
  return /* @__PURE__ */ jsx(
    DynamicSelectionAreaTooltip,
    {
      selectedDatapoints,
      selectState,
      categoriesCount,
      seriesCount,
      axesMetadata,
      selectedSeries,
      selectionDomain,
      selectConfig,
      datapointsMessage
    }
  );
};
const DynamicSelectionAreaTooltip = ({
  selectedDatapoints,
  selectState,
  categoriesCount,
  seriesCount,
  axesMetadata,
  selectedSeries,
  selectionDomain,
  selectConfig,
  datapointsMessage
}) => {
  const intl = useIntl();
  const zoomToSelection = useZoomToSelection();
  const { overlayRef: chartRef } = useOverlay();
  const chartBounds = usePointerAndScrollResponsiveChartBounds(
    chartRef,
    true
  )();
  if (isUndefined(selectState) || !chartBounds) {
    return null;
  }
  const tooltipPosition = getSelectionTooltipPosition(selectState, chartBounds);
  return /* @__PURE__ */ jsx(
    SelectionTooltipWithAsyncActions,
    {
      actions: () => {
        const actions = selectConfig?.actions;
        return actions && (async () => actions(selectedSeries, selectionDomain))().then(
          getCustomContextMenuItems
        );
      },
      selectFinished: selectState.finished,
      header: /* @__PURE__ */ jsx(
        TooltipHeader,
        {
          subheader: intl.formatMessage(datapointsMessage, {
            count: selectedDatapoints,
            seriesCount: seriesCount + categoriesCount
          })
        }
      ),
      position: tooltipPosition,
      zoomEnabled: axesMetadata.isZoomEnabled,
      range: selectState.selectionPercentage || [0, 0],
      children: axesMetadata.isZoomEnabled ? /* @__PURE__ */ jsx(
        OverlayTooltip.Action,
        {
          label: intl.formatMessage(XYCHART_MESSAGES.zoomToSelection),
          icon: /* @__PURE__ */ jsx(ZoomToSelectionIcon, {}),
          onSelect: zoomToSelection,
          trailingIcon: "\u23CE"
        }
      ) : null
    }
  );
};
export {
  XYChartSelectionAreaTooltip
};
