// @dynatrace/strato-components/content/ai-response/tests/vite-helpers.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var vite_helpers_exports = {};
__export(vite_helpers_exports, {
  clearMatchMediaMock: () => clearMatchMediaMock,
  render: () => render,
  setupMatchMediaMock: () => setupMatchMediaMock
});
module.exports = __toCommonJS(vite_helpers_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("@testing-library/react");
var import_react_intl = require("react-intl");
var import_vitest = require("vitest");
const pristineMatchMedia = window.matchMedia;
function setupMatchMediaMock(isMatching = false) {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: import_vitest.vi.fn().mockImplementation((query) => ({
      matches: isMatching,
      media: query,
      onchange: null,
      addListener: import_vitest.vi.fn(),
      // Deprecated
      removeListener: import_vitest.vi.fn(),
      // Deprecated
      addEventListener: import_vitest.vi.fn(),
      removeEventListener: import_vitest.vi.fn(),
      dispatchEvent: import_vitest.vi.fn()
    }))
  });
}
function clearMatchMediaMock() {
  window.matchMedia = pristineMatchMedia;
}
const AllProviders = ({
  children,
  locale = "en",
  timeZone = "UTC"
}) => {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_intl.IntlProvider, { locale, timeZone, children });
};
function render(ui, {
  locale = "en",
  timeZone = "UTC",
  ...renderOptions
} = {}) {
  return (0, import_react.render)(ui, {
    wrapper: ({ children }) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(AllProviders, { locale, timeZone, children }),
    ...renderOptions
  });
}
