// @dynatrace/strato-components/esm/charts/core/components/toolbar/components/ChartDataMenu.js
import { jsx, jsxs } from "react/jsx-runtime";
import { createElement } from "react";
import {
  useCallback,
  useEffect,
  useRef,
  useState
} from "react";
import { useIntl } from "react-intl";
import Spacings from "@dynatrace/strato-design-tokens/spacings";
import { DotMenuIcon, DownloadIcon } from "@dynatrace/strato-icons";
import { AsyncIntents } from "./AsyncIntents.js";
import { triggerButton } from "./ChartDataMenu.sty.js";
import {
  styledToolbarMenuContainer,
  toolbarMenuGroup
} from "./ToolbarMenu.sty.js";
import { Button } from "../../../../../buttons/button/Button.js";
import { Menu } from "../../../../../navigation/menu/Menu.js";
import { useA11yToolbarTabIndex } from "../../../hooks/useA11yToolbarTabIndex.js";
import { useEffectOnValueChange } from "../../../hooks/useEffectOnValueChange.js";
import { TOOLBAR_MESSAGES } from "../messages.js";
import { VerticalDivider } from "./divider.js";
const ChartDataMenu = ({
  onMenuEnter,
  onMenuLeave,
  onMenuOpen,
  menuAlignment = "end",
  isDownloadEnabled,
  handleDownloadDataToolbar,
  intents = [],
  showDivider,
  buttonMode = "standard",
  isDragging,
  intentsAsyncTemplate
}) => {
  const intentButtonRef = useRef(null);
  const menuContentRef = useRef(null);
  const tabIndex = useA11yToolbarTabIndex("intent");
  const intl = useIntl();
  const [isOpen, setIsOpen] = useState(false);
  const isInlineMode = buttonMode === "inline";
  const handleOnMouseEnter = () => {
    onMenuEnter?.();
  };
  useEffectOnValueChange(() => {
    onMenuOpen(isOpen);
  }, [isOpen]);
  useEffect(() => {
    if (tabIndex === 0 && intentButtonRef.current) {
      intentButtonRef.current?.focus();
    }
  }, [intentButtonRef, tabIndex]);
  useEffect(() => {
    if (isDragging !== void 0 && isDragging && isOpen) {
      setIsOpen(false);
    }
  }, [isDragging, isOpen]);
  const handleOnMouseLeave = () => {
    onMenuLeave?.();
  };
  const handleToolbarMenuClick = () => {
    setIsOpen((prevState) => !prevState);
  };
  const handleOnClickOutsideOverlay = useCallback(
    (e) => {
      const menuContainsTarget = menuContentRef.current?.contains(
        e.target
      );
      if (isOpen && !menuContainsTarget) {
        setIsOpen(false);
      }
    },
    [setIsOpen, isOpen, menuContentRef]
  );
  const handleOnKeyUp = useCallback(
    (e) => {
      if (e.key === "Escape") {
        setIsOpen(false);
      }
    },
    [setIsOpen]
  );
  const handleOnKeyDown = (e) => {
    const nativeEvent = e.nativeEvent;
    if (nativeEvent.code === "Enter") {
      setIsOpen((prevState) => !prevState);
    }
  };
  useEffect(() => {
    window.addEventListener("click", handleOnClickOutsideOverlay);
    window.addEventListener("keyup", handleOnKeyUp);
    return () => {
      window.removeEventListener("click", handleOnClickOutsideOverlay);
      window.removeEventListener("keyup", handleOnKeyUp);
    };
  }, [handleOnClickOutsideOverlay, handleOnKeyUp]);
  return /* @__PURE__ */ jsxs(
    "div",
    {
      className: styledToolbarMenuContainer({
        inline: isInlineMode
      }),
      onMouseEnter: handleOnMouseEnter,
      onMouseLeave: handleOnMouseLeave,
      onClick: handleToolbarMenuClick,
      onKeyDown: handleOnKeyDown,
      ref: menuContentRef,
      children: [
        showDivider && /* @__PURE__ */ jsx(VerticalDivider, {}),
        /* @__PURE__ */ jsxs(Menu, { open: isOpen, children: [
          /* @__PURE__ */ jsx(Menu.Trigger, { children: /* @__PURE__ */ jsx(
            Button,
            {
              width: Spacings.Size28,
              className: isInlineMode ? triggerButton : void 0,
              variant: "default",
              size: "condensed",
              ref: intentButtonRef,
              tabIndex,
              children: /* @__PURE__ */ jsx(DotMenuIcon, {})
            }
          ) }),
          /* @__PURE__ */ jsxs(Menu.Content, { alignment: menuAlignment, children: [
            /* @__PURE__ */ jsx(Menu.Group, { className: toolbarMenuGroup, children: isDownloadEnabled && handleDownloadDataToolbar ? /* @__PURE__ */ jsxs(Menu.Item, { onSelect: handleDownloadDataToolbar, children: [
              /* @__PURE__ */ jsx(Menu.Prefix, { children: /* @__PURE__ */ jsx(DownloadIcon, {}) }),
              intl.formatMessage(TOOLBAR_MESSAGES.downloadDataButton)
            ] }) : null }),
            intents.length ? /* @__PURE__ */ jsx(Menu.Group, { className: toolbarMenuGroup, children: intents.map((item, key) => /* @__PURE__ */ createElement(Menu.Intent, { ...item, key: item.children?.toString() || key }, item.icon && /* @__PURE__ */ jsx(Menu.Prefix, { children: item.icon }), item.children)) }) : null,
            /* @__PURE__ */ jsx(
              AsyncIntents,
              {
                intentsAsyncTemplate,
                shouldRun: isOpen
              }
            )
          ] })
        ] })
      ]
    }
  );
};
export {
  ChartDataMenu
};
