// @dynatrace/strato-components/esm/charts/core/components/toolbar/components/Control.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import {
  forwardRef,
  useEffect,
  useRef
} from "react";
import {
  buttonStyle,
  menuItemStyles,
  styledMenuItemContent,
  styledShortcut,
  styledText
} from "./Control.sty.js";
import { Button } from "../../../../../buttons/button/Button.js";
import { KeyboardShortcutTooltip } from "../../../../../content/keyboard-shortcut-tooltip/KeyboardShortcutTooltip.js";
import { useMergeRefs } from "../../../../../core/hooks/useMergeRefs.js";
import { Divider } from "../../../../../layouts/divider/Divider.js";
import { Menu } from "../../../../../navigation/menu/Menu.js";
import { Tooltip } from "../../../../../overlays/tooltip/Tooltip.js";
import {
  styledHorizontalDivider,
  styledVerticalDivider
} from "../ChartToolbarRenderer.sty.js";
import { useToolbarContext } from "../hooks/useToolbarContext.js";
const Control = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const {
    prefixIcon,
    tooltipText,
    testId,
    selected = false,
    onClick,
    onKeyDown,
    tabIndex,
    disabled = false,
    hideTooltip = false,
    tooltipKeyShortcut,
    placement = "auto",
    showWhenCollapsed = false,
    divider = "none",
    className: consumerClassName,
    style: consumerStyle,
    "data-dt-name": dtName,
    "data-dt-properties": dtProperties
  } = props;
  const { isMenuElement, isExpanded, didDrag, resetDidDrag } = useToolbarContext();
  const localRef = useRef(null);
  const rootRef = useMergeRefs([
    forwardedRef,
    localRef
  ]);
  useEffect(() => {
    if (tabIndex === 0) {
      localRef.current?.focus();
    }
  }, [tabIndex]);
  const handleClick = (event) => {
    event.stopPropagation();
    if (didDrag) {
      resetDidDrag?.();
      return;
    }
    if (!disabled) {
      onClick(event);
    }
  };
  const handleSelect = (event) => {
    event.stopPropagation();
    if (!disabled) {
      onClick(event);
    }
  };
  const handleKeyDown = (event) => {
    if (!disabled) {
      onKeyDown?.(event);
    }
  };
  const styledButton = /* @__PURE__ */ jsx(
    Button,
    {
      ref: rootRef,
      className: clsx(buttonStyle({ disabled, selected }), consumerClassName),
      style: consumerStyle,
      variant: "default",
      tabIndex,
      "data-testid": testId,
      "aria-label": tooltipText,
      onKeyDown: handleKeyDown,
      onMouseUp: handleClick,
      size: "condensed",
      "data-dt-children-name": dtName,
      "data-dt-properties": dtProperties,
      children: /* @__PURE__ */ jsx(Button.Prefix, { children: prefixIcon })
    }
  );
  if (isMenuElement) {
    if (placement === "toolbar" || placement === "auto" && isExpanded) {
      return null;
    }
    const horizontalDivider = /* @__PURE__ */ jsx(
      Divider,
      {
        className: styledHorizontalDivider,
        orientation: "horizontal",
        as: "hr"
      }
    );
    return /* @__PURE__ */ jsxs(Fragment, { children: [
      divider === "left" && horizontalDivider,
      /* @__PURE__ */ jsxs(
        Menu.Item,
        {
          className: menuItemStyles({ disabled, selected }),
          onSelect: handleSelect,
          onKeyDown: handleKeyDown,
          "data-dt-children-name": dtName,
          "data-dt-properties": dtProperties,
          children: [
            /* @__PURE__ */ jsx(Menu.Prefix, { children: prefixIcon }),
            /* @__PURE__ */ jsxs("div", { className: styledMenuItemContent, children: [
              /* @__PURE__ */ jsx("span", { className: styledText, children: tooltipText }),
              /* @__PURE__ */ jsxs("span", { className: styledShortcut({ disabled }), children: [
                "\xA0",
                tooltipKeyShortcut
              ] })
            ] })
          ]
        }
      ),
      divider === "right" && horizontalDivider
    ] });
  }
  if (placement === "menu" || !isExpanded && !showWhenCollapsed) {
    return null;
  }
  const verticalDivider = /* @__PURE__ */ jsx(Divider, { className: styledVerticalDivider, orientation: "vertical", as: "hr" });
  let computedControlButton = styledButton;
  if (tooltipText) {
    if (tooltipKeyShortcut) {
      computedControlButton = /* @__PURE__ */ jsx(
        KeyboardShortcutTooltip,
        {
          keys: tooltipKeyShortcut,
          text: tooltipText,
          disabled: hideTooltip,
          shortcutDisabled: disabled,
          children: styledButton
        }
      );
    } else {
      computedControlButton = /* @__PURE__ */ jsx(Tooltip, { text: tooltipText, disabled: hideTooltip, children: styledButton });
    }
  }
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    (divider === "left" || placement === "toolbar" && !isExpanded) && verticalDivider,
    computedControlButton,
    divider === "right" && verticalDivider
  ] });
});
Control.displayName = "ChartToolbar.Control";
export {
  Control
};
