// @dynatrace/strato-components/esm/charts/core/components/toolbar/components/ToolbarMenu.js
import { jsx, jsxs } from "react/jsx-runtime";
import {
  Children,
  cloneElement,
  Fragment,
  isValidElement,
  useEffect,
  useMemo,
  useRef,
  useState
} from "react";
import { useIntl } from "react-intl";
import { ChevronDownSmallIcon, CrosshairIcon } from "@dynatrace/strato-icons";
import { Button } from "../../../../../buttons/button/Button.js";
import { Menu } from "../../../../../navigation/menu/Menu.js";
import { Tooltip } from "../../../../../overlays/tooltip/Tooltip.js";
import { useA11yToolbarTabIndex } from "../../../hooks/useA11yToolbarTabIndex.js";
import { useToolbarContext } from "../hooks/useToolbarContext.js";
import { TOOLBAR_MESSAGES } from "../messages.js";
import {
  styledToolbarMenuContainer,
  toolbarMenuGroup
} from "./ToolbarMenu.sty.js";
function renderMenuItems(children) {
  return Children.map(children, (child) => {
    if (isValidElement(child)) {
      if (child.type === Fragment) {
        return renderMenuItems(child.props.children);
      }
      return cloneElement(child, {
        ...child.props
      });
    }
    return child;
  });
}
const ToolbarMenu = (props) => {
  const {
    children,
    onMenuOpen,
    isOpen,
    activeModeIcon,
    triggerLabel,
    onMenuEnter,
    onMenuLeave,
    menuAlignment = "start",
    onCloseAutoFocus,
    "data-dt-name": dtName,
    "data-dt-properties": dtProperties
  } = props;
  const [isHoverChartToolsIcon, setIsHoverChartToolsIcon] = useState(false);
  const menuTriggerButtonRef = useRef(null);
  const tabIndex = useA11yToolbarTabIndex("explore");
  const intl = useIntl();
  const { isExpanded } = useToolbarContext();
  const menuItems = useMemo(() => renderMenuItems(children), [children]);
  useEffect(() => {
    if (tabIndex === 0 && menuTriggerButtonRef.current) {
      menuTriggerButtonRef.current?.focus();
    }
  }, [menuTriggerButtonRef, tabIndex]);
  const chartActionsButton = useMemo(() => {
    const tooltipText = isExpanded ? triggerLabel ?? intl.formatMessage(TOOLBAR_MESSAGES.changeMode) : intl.formatMessage(TOOLBAR_MESSAGES.chartTools);
    const handleMenuOpenChange = (isOpen2) => setIsHoverChartToolsIcon(isOpen2);
    return /* @__PURE__ */ jsx(
      Tooltip,
      {
        text: tooltipText,
        open: isHoverChartToolsIcon,
        onOpenChange: handleMenuOpenChange,
        children: /* @__PURE__ */ jsx(Menu.Trigger, { children: /* @__PURE__ */ jsxs(
          Button,
          {
            variant: "default",
            size: "condensed",
            ref: menuTriggerButtonRef,
            tabIndex,
            "data-dt-name": dtName,
            "data-dt-properties": dtProperties,
            children: [
              /* @__PURE__ */ jsx(Button.Prefix, { children: activeModeIcon ?? /* @__PURE__ */ jsx(CrosshairIcon, {}) }),
              /* @__PURE__ */ jsx(Button.Suffix, { children: /* @__PURE__ */ jsx(ChevronDownSmallIcon, {}) })
            ]
          }
        ) })
      }
    );
  }, [
    isHoverChartToolsIcon,
    isExpanded,
    activeModeIcon,
    triggerLabel,
    intl,
    tabIndex,
    dtName,
    dtProperties
  ]);
  const handleOnMouseEnter = () => {
    onMenuEnter?.();
  };
  const handleOnMouseLeave = () => {
    onMenuLeave?.();
  };
  return /* @__PURE__ */ jsx(
    "div",
    {
      className: styledToolbarMenuContainer({
        inline: props.toolbarDisplayMode === "inline"
      }),
      onMouseEnter: handleOnMouseEnter,
      onMouseLeave: handleOnMouseLeave,
      children: /* @__PURE__ */ jsxs(Menu, { onOpenChange: onMenuOpen, open: isOpen, children: [
        chartActionsButton,
        /* @__PURE__ */ jsx(
          Menu.Content,
          {
            alignment: menuAlignment,
            onCloseAutoFocus,
            children: /* @__PURE__ */ jsx(Menu.Group, { className: toolbarMenuGroup, children: menuItems })
          }
        )
      ] })
    }
  );
};
export {
  ToolbarMenu
};
