// @dynatrace/strato-components/layouts/title-bar/TitleBar.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TitleBar_exports = {};
__export(TitleBar_exports, {
  TitleBar: () => TitleBar
});
module.exports = __toCommonJS(TitleBar_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_spacings = __toESM(require("@dynatrace/strato-design-tokens/spacings"));
var import_Action = require("./Action.js");
var import_Navigation = require("./Navigation.js");
var import_Prefix = require("./Prefix.js");
var import_Subtitle = require("./Subtitle.js");
var import_Suffix = require("./Suffix.js");
var import_Title = require("./Title.js");
var import_TitleBar_css = require("./TitleBar.sty.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useSlots = require("../../core/hooks/useSlots.js");
var import_can_receive_default_text_format = require("../../core/utils/can-receive-default-text-format.js");
var import_Divider = require("../divider/Divider.js");
var import_Flex = require("../flex/Flex.js");
const COMPONENT_NAME = "TitleBar";
const TitleBarRoot = /* @__PURE__ */ (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const {
      children,
      showDivider = true,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...remainingProps
    } = props;
    const [slots] = (0, import_useSlots.useSlots)(children, {
      title: import_Title.Title,
      subtitle: import_Subtitle.Subtitle,
      navigation: import_Navigation.Navigation,
      suffix: import_Suffix.Suffix,
      prefix: import_Prefix.Prefix,
      action: import_Action.Action
    });
    const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
    const titleIsText = !!slots.title && (0, import_can_receive_default_text_format.canReceiveDefaultTextFormat)(slots.title.props.children);
    const subtitleIsText = !!slots.subtitle && (0, import_can_receive_default_text_format.canReceiveDefaultTextFormat)(slots.subtitle?.props.children);
    const titleAndSubtitleIsText = titleIsText && !slots.subtitle || titleIsText && subtitleIsText;
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      import_Flex.Flex,
      {
        ref: forwardedRef,
        ...restProps,
        ...behavioralTrackingProps,
        width: "100%",
        flexDirection: "column",
        gap: 0,
        className: consumerClassName,
        style: consumerStyle,
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        "data-dt-component": COMPONENT_NAME,
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { flexDirection: "column", gap: 8, children: [
            slots.navigation && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { width: "100%", justifyContent: "space-between", gap: 8, children: [
              slots.navigation,
              slots.action
            ] }),
            /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { width: "100%", justifyContent: "space-between", gap: 8, children: [
              /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { flexWrap: "wrap", gap: 8, width: "100%", minWidth: 0, children: [
                slots.prefix,
                (slots.title || slots.subtitle) && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
                  "div",
                  {
                    className: (0, import_TitleBar_css.titleSubtitleWrapperCSS)({
                      childrenAreText: titleAndSubtitleIsText
                    }),
                    children: [
                      slots.title,
                      slots.subtitle
                    ]
                  }
                ),
                slots.suffix
              ] }),
              !slots.navigation && slots.action
            ] })
          ] }),
          showDivider && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Divider.Divider, { flexItem: true, style: { marginTop: import_spacings.default.Size8 } })
        ]
      }
    );
  }
);
TitleBarRoot.displayName = "TitleBar";
const TitleBar = Object.assign(TitleBarRoot, {
  Title: import_Title.Title,
  Subtitle: import_Subtitle.Subtitle,
  Suffix: import_Suffix.Suffix,
  Navigation: import_Navigation.Navigation,
  Prefix: import_Prefix.Prefix,
  Action: import_Action.Action
});
