// @dynatrace/strato-components/esm/charts/core/components/tooltip/section-components/atoms/TooltipText.js
import { jsx } from "react/jsx-runtime";
import { useRef } from "react";
import { useIntl } from "react-intl";
import { OverlayStackingProvider } from "../../../../../../core/providers/OverlayStackingProvider.js";
import { Tooltip } from "../../../../../../overlays/tooltip/Tooltip.js";
import { DEFAULT_TRUNCATION_MODE } from "../../../../constants.js";
import { useTruncationDetector } from "../../../../hooks/useTruncationDetector.js";
import { ChartTextEllipsis } from "../../../text-ellipsis/ChartTextEllipsis.js";
import { PrimaryAtom } from "../../../tooltip-atoms/components/PrimaryAtom.js";
import { SecondaryAtom } from "../../../tooltip-atoms/components/SecondaryAtom.js";
const TooltipText = (props) => {
  const {
    children: text,
    variant = "primary",
    truncationMode = DEFAULT_TRUNCATION_MODE,
    ...rest
  } = props;
  const textRef = useRef(null);
  const isTruncated = useTruncationDetector(textRef, text);
  const intl = useIntl();
  return /* @__PURE__ */ jsx(OverlayStackingProvider, { children: /* @__PURE__ */ jsx(Tooltip, { text, disabled: !isTruncated, children: variant === "primary" ? /* @__PURE__ */ jsx(PrimaryAtom, { ...rest, children: /* @__PURE__ */ jsx(
    ChartTextEllipsis,
    {
      ref: textRef,
      truncationMode,
      "aria-label": intl.formatMessage({
        defaultMessage: "Primary series name",
        id: "OwC/n+tx4kFWA3l3"
      }),
      children: text
    }
  ) }) : /* @__PURE__ */ jsx(SecondaryAtom, { ...rest, children: /* @__PURE__ */ jsx(
    ChartTextEllipsis,
    {
      ref: textRef,
      truncationMode,
      "aria-label": intl.formatMessage({
        defaultMessage: "Secondary series name",
        id: "QbeiZKf6kqgyTp3m"
      }),
      children: text
    }
  ) }) }) });
};
TooltipText["displayName"] = "TooltipText";
export {
  TooltipText
};
