// @dynatrace/strato-components/esm/filters/timeframe-selector/components/trigger/Trigger.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import {
  Children,
  useMemo
} from "react";
import { isElement } from "react-is";
import { ChevronDownSmallIcon } from "@dynatrace/strato-icons";
import { isTimeframeSelectorDisplayValueElement } from "./DisplayValue.js";
import * as styles from "./Trigger.sty.js";
import { Button } from "../../../../buttons/button/Button.js";
import { isVisualRefreshEnabled } from "../../../../core/utils/is-visual-refresh-enabled.js";
import { mergeProps as _mergeProps } from "../../../../core/utils/merge-props.js";
import { useTimeframeSelectorContext } from "../../hooks/useTimeframeSelectorContext.js";
function isTimeframeSelectorTriggerElement(child) {
  return isElement(child) && child.type === Trigger;
}
function getDisplayValueSlot(children) {
  let displayValue = null;
  Children.forEach(children, (child) => {
    if (isTimeframeSelectorDisplayValueElement(child)) {
      displayValue = displayValue || [];
      displayValue.push(child);
    }
  });
  return displayValue;
}
const Trigger = (props) => {
  const {
    children,
    placeholder,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    ...remainingProps
  } = props;
  const displayValueSlot = useMemo(
    () => getDisplayValueSlot(children),
    [children]
  );
  const {
    triggerProps,
    displayValue: contextDisplayValue,
    details,
    readOnly,
    showError,
    disabled
  } = useTimeframeSelectorContext();
  const isCritical = isVisualRefreshEnabled() ? showError && !disabled : showError;
  const color = isCritical ? "critical" : "neutral";
  const displayValue = displayValueSlot ?? contextDisplayValue;
  const placeholderIfEmpty = placeholder ?? contextDisplayValue;
  return /* @__PURE__ */ jsxs(
    Button,
    {
      ..._mergeProps(triggerProps, remainingProps),
      "data-testid": dataTestId,
      variant: "emphasized",
      color,
      textAlign: "start",
      style: consumerStyle,
      className: clsx(consumerClassName, {
        [styles.readOnly]: readOnly,
        [styles.readOnlyColor[color]]: readOnly
      }),
      children: [
        /* @__PURE__ */ jsx(Button.Suffix, { children: /* @__PURE__ */ jsx(
          ChevronDownSmallIcon,
          {
            className: clsx(
              styles.suffix,
              triggerProps?.["aria-expanded"] && styles.suffixOpen
            )
          }
        ) }),
        details ? displayValue : placeholderIfEmpty
      ]
    }
  );
};
Trigger.displayName = "TimeframeSelector.Trigger";
export {
  Trigger,
  isTimeframeSelectorTriggerElement
};
