// @dynatrace/strato-components/esm/content/chip-group/state/chip-group-state-reducer.js
var ChipGroupActions = /* @__PURE__ */ ((ChipGroupActions2) => {
  ChipGroupActions2["ADDCHIP"] = "add";
  ChipGroupActions2["UPDATECHIP"] = "update";
  ChipGroupActions2["REMOVECHIP"] = "remove";
  ChipGroupActions2["ADDCONTROL"] = "add-control";
  ChipGroupActions2["UPDATECONTROL"] = "update-control";
  ChipGroupActions2["REMOVECONTROL"] = "remove-control";
  return ChipGroupActions2;
})(ChipGroupActions || {});
function chipGroupReducer(state, action) {
  const { type, payload } = action;
  const newState = { ...state };
  if (type === "add" /* ADDCHIP */) {
    const { width } = payload;
    newState.chips = { ...newState.chips, [payload.uniqueId]: { width } };
    return newState;
  }
  if (type === "update" /* UPDATECHIP */) {
    const { uniqueId: _, ...chipState } = payload;
    newState.chips = {
      ...newState.chips,
      [payload.uniqueId]: {
        ...newState.chips[payload.uniqueId],
        ...chipState
      }
    };
    return newState;
  }
  if (type === "remove" /* REMOVECHIP */) {
    delete newState.chips[payload.uniqueId];
    return newState;
  }
  if (type === "add-control" /* ADDCONTROL */) {
    const { width } = payload;
    newState.control = { width, show: "auto" };
    return newState;
  }
  if (type === "update-control" /* UPDATECONTROL */) {
    const { width, show } = payload;
    newState.control = { width, show };
    return newState;
  }
  if (type === "remove-control" /* REMOVECONTROL */) {
    newState.control = void 0;
    return newState;
  }
  return state;
}
export {
  ChipGroupActions,
  chipGroupReducer
};
