// @dynatrace/strato-components/tables/DataTable/features/UserActions/FloatingMenu.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var FloatingMenu_exports = {};
__export(FloatingMenu_exports, {
  CellActionFloatingMenu: () => CellActionFloatingMenu
});
module.exports = __toCommonJS(FloatingMenu_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_dom = require("react-dom");
var styles = __toESM(require("./UserActions.sty.js"));
var import_Menu = require("../../../../navigation/menu/Menu.js");
var import_DataTableContext = require("../../contexts/DataTableContext.js");
function CellActionFloatingMenu() {
  const table = (0, import_DataTableContext.useDataTable)();
  const { type, headerId, rowId, columnId, x, y } = table.getState()?.floatingMenuInfo ?? {
    type: null,
    headerId: null,
    rowId: null,
    columnId: null
  };
  const actions = (0, import_react.useMemo)(() => {
    if (type === "headerAction" && columnId !== null) {
      return table.getUserActionsForColumn(columnId);
    }
    if (type === "cellAction" && columnId !== null && rowId !== null) {
      return table.getUserActionsForCell(columnId, rowId);
    }
    return null;
  }, [rowId, columnId, type, table]);
  const menuOpen = type !== null && columnId !== null && actions !== null;
  const { header, column, cell } = (0, import_react.useMemo)(() => {
    if (menuOpen === false) {
      return { column: void 0, cell: void 0 };
    }
    const header2 = headerId ? table.getHeader(headerId) : void 0;
    const column2 = table.getColumn(columnId);
    let cell2 = void 0;
    if (rowId) {
      const rowCells = table.getRow(rowId).getAllCells();
      const columnCell = rowCells.find((cell3) => cell3.column.id === columnId);
      cell2 = columnCell;
    }
    return { header: header2, column: column2, cell: cell2 };
  }, [columnId, headerId, menuOpen, rowId, table]);
  if (menuOpen === false) {
    return;
  }
  return (0, import_react_dom.createPortal)(
    /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      import_Menu.Menu,
      {
        open: menuOpen,
        onOpenChange: (isOpen) => {
          if (isOpen === false) {
            table.setState((state) => {
              const newState = { ...state };
              newState.floatingMenuInfo = {
                type: null,
                headerId: null,
                rowId: null,
                columnId: null,
                x: 0,
                y: 0
              };
              newState.openMenuRowId = null;
              return newState;
            });
          }
        },
        modal: false,
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Trigger, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "button",
            {
              disabled: true,
              type: "button",
              className: styles.floatingMenuButton,
              style: {
                top: `${y}px`,
                left: `${x}px`
              },
              "data-trigger": "context-action"
            }
          ) }),
          menuOpen === true && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_DataTableContext.DataTableHeaderContextProvider, { value: header, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_DataTableContext.DataTableColumnContextProvider, { value: column, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_DataTableContext.DataTableCellContextProvider, { value: cell, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Content, { children: import_react.Children.map(actions, (child) => child) }) }) }) })
        ]
      }
    ),
    document.body
  );
}
