// @dynatrace/strato-components/tables/DataTable/features/UserActions/InlineUserActionsMenu.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var InlineUserActionsMenu_exports = {};
__export(InlineUserActionsMenu_exports, {
  InlineUserActionsMenu: () => InlineUserActionsMenu
});
module.exports = __toCommonJS(InlineUserActionsMenu_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_useCellActions = require("./useCellActions.js");
var styles = __toESM(require("./UserActions.sty.js"));
var import_Menu = require("../../../../navigation/menu/Menu.js");
var import_constants = require("../../constants.js");
var import_DataTableContext = require("../../contexts/DataTableContext.js");
function InlineUserActionsMenu(props) {
  const {
    columnId,
    headerGroupOrRowId,
    type,
    "aria-label": ariaLabel,
    "data-header-row": dataHeaderRow,
    "data-column": dataColumn,
    hovered,
    onClose
  } = props;
  const [menuOpen, setMenuOpen] = (0, import_react.useState)(false);
  const table = (0, import_DataTableContext.useDataTable)();
  const column = table.getColumn(columnId);
  const actions = (0, import_useCellActions.useCellAndColumnActions)({
    enabled: menuOpen,
    headerGroupOrRowId,
    columnId,
    type
  });
  const floatingMenuInfo = table.getState().floatingMenuInfo;
  const isFloatingMenuOpenForThis = floatingMenuInfo?.columnId === columnId && (type === "cell" && floatingMenuInfo?.rowId === headerGroupOrRowId || type === "header" && floatingMenuInfo?.type === "headerAction");
  if (isFloatingMenuOpenForThis) {
    return void 0;
  }
  function getSubrowIndentation() {
    if (!table.options.enableExpanding) {
      return void 0;
    }
    if (table.getSubRowIndicatorColumnId() !== columnId) {
      return void 0;
    }
    const rowDensity = table.getVisualVariantOptions().rowDensity;
    if (type === "header") {
      if (column?.columnDef.meta?.hasCustomHeaderRenderer) {
        return void 0;
      }
      return import_constants.DATA_TABLE_SUB_ROW_EXPANDER_ICON_SIZE + import_constants.DATA_TABLE_CELL_HORIZONTAL_PADDING_MAP[rowDensity];
    }
    const row = table.getRow(headerGroupOrRowId);
    return row.depth * import_constants.DATA_TABLE_SUB_ROW_INDENT_GRID + import_constants.DATA_TABLE_SUB_ROW_EXPANDER_ICON_SIZE + import_constants.DATA_TABLE_CELL_HORIZONTAL_PADDING_MAP[rowDensity];
  }
  const triggerProps = {
    type: "button",
    className: styles.actionButton,
    style: {
      display: menuOpen ? "inline-block" : void 0,
      marginLeft: getSubrowIndentation()
    },
    "aria-label": ariaLabel,
    "data-header-row": dataHeaderRow,
    "data-column": dataColumn,
    "data-trigger": "useraction",
    children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_strato_icons.DotMenuIcon,
      {
        style: {
          width: import_constants.DATA_TABLE_DEFAULT_ICON_SIZE,
          height: import_constants.DATA_TABLE_DEFAULT_ICON_SIZE
        }
      }
    )
  };
  if (!menuOpen) {
    if (hovered === false) {
      return;
    }
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "button",
      {
        ...triggerProps,
        onClick: (event) => {
          event.preventDefault();
          event.stopPropagation();
          setMenuOpen(true);
          if (type === "cell") {
            table.setOpenMenuRowId(headerGroupOrRowId);
          }
        }
      }
    );
  }
  const alignment = column?.columnDef.alignment?.includes("right") ? "start" : "end";
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_Menu.Menu,
    {
      open: menuOpen,
      onOpenChange: (open) => {
        if (!open && onClose) {
          onClose();
        }
        if (!open && type === "cell") {
          table.setOpenMenuRowId(null);
        }
        setMenuOpen(open);
      },
      modal: false,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Trigger, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          "button",
          {
            ...triggerProps,
            onClick: (evt) => {
              evt.stopPropagation();
            }
          }
        ) }),
        menuOpen === true ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_DataTableContext.DataTableColumnContextProvider, { value: column, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Content, { alignment, children: import_react.Children.map(actions, (child) => child) }) }) : null
      ]
    }
  );
}
