// @dynatrace/strato-components/tables/DataTable/features/UserActions/RowActions.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var RowActions_exports = {};
__export(RowActions_exports, {
  DataTableRowActions: () => DataTableRowActions,
  ROW_ACTIONS_COLUMN: () => ROW_ACTIONS_COLUMN,
  isDataTableRowActions: () => isDataTableRowActions
});
module.exports = __toCommonJS(RowActions_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_react_is = require("react-is");
var import_strato_icons = require("@dynatrace/strato-icons");
var styles = __toESM(require("./RowActions.sty.js"));
var import_Button = require("../../../../buttons/button/Button.js");
var import_Menu = require("../../../../navigation/menu/Menu.js");
var import_Trigger = require("../../../../navigation/menu/Trigger.js");
var import_DataTableDefaultCell = require("../../components/DataTableDefaultCell.js");
var import_constants = require("../../constants.js");
var import_DataTableContext = require("../../contexts/DataTableContext.js");
var import_messages = require("../../messages.js");
function DataTableRowActions(props) {
  return null;
}
function isDataTableRowActions(child) {
  if ((0, import_react_is.isElement)(child) && child.type === DataTableRowActions) {
    return true;
  }
  return false;
}
const DataTableInternalRowActions = (0, import_react.memo)(
  (props) => {
    const table = (0, import_DataTableContext.useDataTable)();
    const intl = (0, import_react_intl.useIntl)();
    const density = table.getVisualVariantOptions().rowDensity === "condensed" ? "condensed" : "default";
    const ariaLabel = intl.formatMessage(import_messages.dataTableMessages.rowActionsTrigger);
    const [open, setOpen] = (0, import_react.useState)(false);
    const handleOpenChange = (nextOpen) => {
      setOpen(nextOpen);
      table.setOpenMenuRowId(nextOpen ? props.rowId : null);
    };
    const { menu, freeformSlot } = (0, import_react.useMemo)(() => {
      const slots = {
        menu: null,
        freeformSlot: []
      };
      const childrenArray = import_react.Children.toArray(props.children);
      for (const child of childrenArray) {
        if ((0, import_react.isValidElement)(child) && child.type === import_Menu.Menu) {
          slots.menu = child;
          continue;
        }
        slots.freeformSlot.push(child);
      }
      return slots;
    }, [props.children]);
    if (!menu) {
      return freeformSlot;
    }
    if (open === false) {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
        freeformSlot,
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          "span",
          {
            onClick: () => handleOpenChange(true),
            style: { display: "contents" },
            children: getMenuTrigger(menu, density, ariaLabel)
          }
        )
      ] });
    }
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
      freeformSlot,
      /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
        import_Menu.Menu,
        {
          open,
          onOpenChange: handleOpenChange,
          modal: false,
          ...menu.props,
          children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Trigger, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button, { variant: "default", size: density, "aria-label": ariaLabel, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.DotMenuIcon, {}) }) }) }),
            menu.props.children
          ]
        }
      )
    ] });
  }
);
function getMenuTrigger(menu, size, ariaLabel) {
  const menuChildren = import_react.Children.toArray(menu.props.children);
  const customMenuTrigger = menuChildren.find(import_Trigger.isTrigger);
  if (customMenuTrigger) {
    return customMenuTrigger.props.children;
  }
  return getDefaultMenuTrigger(size, ariaLabel);
}
function getDefaultMenuTrigger(size, ariaLabel) {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button, { variant: "default", size, "aria-label": ariaLabel, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.DotMenuIcon, {}) }) });
}
function isFragment(child) {
  return child.type === import_react.Fragment;
}
const ROW_ACTIONS_COLUMN = {
  id: import_constants.DATA_TABLE_ROW_ACTIONS_COLUMN_ID,
  accessorFn: () => "",
  header: "",
  width: "content",
  enableSorting: false,
  enableUserActions: false,
  enableHiding: false,
  enableResizing: false,
  meta: {
    builtin: true,
    builtinPlacement: "suffix",
    hasCustomCellRenderer: true,
    hasCustomHeaderRenderer: true
  },
  cell: ({ row, table }) => {
    if (table.options.rowActions === void 0) {
      return null;
    }
    const actions = table.options.rowActions(row.original, {
      rowDensity: table.getVisualVariantOptions().rowDensity,
      rowIndex: row.index,
      rowId: row.id
    });
    if ((0, import_react_is.isElement)(actions)) {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_DataTableDefaultCell.DataTableDefaultCell, { className: styles.rowActions, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(DataTableInternalRowActions, { rowId: row.id, children: isFragment(actions) ? actions.props.children : actions }) });
    }
  }
};
