// @dynatrace/strato-components/tables/DataTable/features/UserActions/UserActionsMenu.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var UserActionsMenu_exports = {};
__export(UserActionsMenu_exports, {
  ColumnOrderAction: () => ColumnOrderAction,
  DataTableActionsMenu: () => DataTableActionsMenu,
  HideColumnAction: () => HideColumnAction
});
module.exports = __toCommonJS(UserActionsMenu_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_useCopyToClipboard = require("../../../../core/utils/useCopyToClipboard.js");
var import_Group = require("../../../../navigation/menu/Group.js");
var import_Intent = require("../../../../navigation/menu/Intent.js");
var import_Item = require("../../../../navigation/menu/Item.js");
var import_Label = require("../../../../navigation/menu/Label.js");
var import_Link = require("../../../../navigation/menu/Link.js");
var import_Prefix = require("../../../../navigation/menu/Prefix.js");
var import_Sub = require("../../../../navigation/menu/Sub.js");
var import_SubContent = require("../../../../navigation/menu/SubContent.js");
var import_SubTrigger = require("../../../../navigation/menu/SubTrigger.js");
var import_Suffix = require("../../../../navigation/menu/Suffix.js");
var import_DataTableContext = require("../../contexts/DataTableContext.js");
var import_messages = require("../../messages.js");
const CopyItem = ({ value, onSelect }) => {
  const intl = (0, import_react_intl.useIntl)();
  const [copyToClipboard] = (0, import_useCopyToClipboard.useCopyToClipboard)(value);
  const header = (0, import_react.useContext)(import_DataTableContext.DataTableHeaderContext);
  const isColumnAction = header !== null;
  const dataDtName = isColumnAction ? "ColumnActionCopyItem" : "CellActionCopyItem";
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    DataTableActionsMenu.Item,
    {
      onSelect: () => {
        copyToClipboard();
        onSelect?.();
      },
      "data-dt-name": dataDtName,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(DataTableActionsMenu.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.CopyIcon, {}) }),
        intl.formatMessage(import_messages.dataTableMessages.copyValueCellAction)
      ]
    }
  );
};
const LineWrapAction = ({
  onSelect
}) => {
  const intl = (0, import_react_intl.useIntl)();
  const column = (0, import_DataTableContext.useDataTableColumn)();
  const isLineWrapped = column?.getIsLineWrapped();
  const lineWrapDtProperties = `state:${isLineWrapped ? "unwrap" : "wrap"}`;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    DataTableActionsMenu.Item,
    {
      onSelect: () => {
        column.toggleLineWrap();
        onSelect?.();
      },
      "data-dt-name": "ColumnActionLineWrap",
      "data-dt-properties": lineWrapDtProperties,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(DataTableActionsMenu.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.WrapTextIcon, {}) }),
        isLineWrapped ? intl.formatMessage(import_messages.dataTableMessages.disableLineWrapLabel) : intl.formatMessage(import_messages.dataTableMessages.enableLineWrapLabel)
      ]
    }
  );
};
const HideColumnAction = ({
  onSelect
}) => {
  const intl = (0, import_react_intl.useIntl)();
  const table = (0, import_DataTableContext.useDataTable)();
  const currentVisibility = table.getState().columnVisibility;
  const column = (0, import_DataTableContext.useDataTableColumn)();
  const childColumns = column.columns;
  const handleSelect = () => {
    if (childColumns.length === 0) {
      column.toggleVisibility();
    } else {
      const updatedChildrenVisibility = childColumns.reduce(
        (updatedVisibility, childColumn) => {
          return {
            ...updatedVisibility,
            [childColumn.id]: false
          };
        },
        {}
      );
      table.setColumnVisibility({
        ...currentVisibility,
        ...updatedChildrenVisibility
      });
    }
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    DataTableActionsMenu.Item,
    {
      onSelect: () => {
        handleSelect();
        onSelect?.();
      },
      disabled: !column.getCanHide(),
      "data-dt-name": "ColumnActionHide",
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(DataTableActionsMenu.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ViewOffIcon, {}) }),
        intl.formatMessage(import_messages.dataTableMessages.hideColumnLabel)
      ]
    }
  );
};
const ColumnOrderAction = ({
  onSelect
}) => {
  const intl = (0, import_react_intl.useIntl)();
  const column = (0, import_DataTableContext.useDataTableColumn)();
  const header = (0, import_DataTableContext.useDataTableHeader)();
  const isPinnedColumn = !!column.getIsPinned();
  const table = (0, import_DataTableContext.useDataTable)();
  if (!table.options.enableColumnOrdering) {
    return null;
  }
  const orderActions = () => /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      DataTableActionsMenu.Item,
      {
        onSelect: () => {
          column.moveColumnLeft();
          onSelect?.();
        },
        disabled: !column.getCanMoveColumnLeft(),
        "data-dt-name": "ColumnActionMoveLeft",
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(DataTableActionsMenu.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ArrowLeftIcon, {}) }),
          intl.formatMessage(import_messages.dataTableMessages.moveColumnLeftLabel)
        ]
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      DataTableActionsMenu.Item,
      {
        onSelect: () => {
          column.moveColumnRight();
          onSelect?.();
        },
        disabled: !column.getCanMoveColumnRight(),
        "data-dt-name": "ColumnActionMoveRight",
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(DataTableActionsMenu.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ArrowRightIcon, {}) }),
          intl.formatMessage(import_messages.dataTableMessages.moveColumnRightLabel)
        ]
      }
    )
  ] });
  if (isPinnedColumn) {
    if (header.id.includes("center")) {
      return orderActions();
    }
    return null;
  }
  return orderActions();
};
const ColumnPinningAction = ({
  onSelect
}) => {
  const intl = (0, import_react_intl.useIntl)();
  const table = (0, import_DataTableContext.useDataTable)();
  const header = (0, import_DataTableContext.useDataTableHeader)();
  if (!table.options.enableColumnPinning) {
    return null;
  }
  const pinnedState = header.getIsPinned();
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    pinnedState !== "left" ? /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      DataTableActionsMenu.Item,
      {
        onSelect: () => {
          header.pin("left");
          onSelect?.();
        },
        "data-dt-name": "ColumnActionPinLeft",
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(DataTableActionsMenu.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.PinIcon, {}) }),
          intl.formatMessage(import_messages.dataTableMessages.pinColumnLeftLabel)
        ]
      }
    ) : null,
    pinnedState !== "right" ? /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      DataTableActionsMenu.Item,
      {
        onSelect: () => {
          header.pin("right");
          onSelect?.();
        },
        "data-dt-name": "ColumnActionPinRight",
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(DataTableActionsMenu.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.PinIcon, {}) }),
          intl.formatMessage(import_messages.dataTableMessages.pinColumnRightLabel)
        ]
      }
    ) : null,
    pinnedState !== false ? /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      DataTableActionsMenu.Item,
      {
        onSelect: () => {
          header.pin(false);
          onSelect?.();
        },
        "data-dt-name": "ColumnActionUnpin",
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(DataTableActionsMenu.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.PinOffIcon, {}) }),
          intl.formatMessage(import_messages.dataTableMessages.unpinColumnLabel)
        ]
      }
    ) : null
  ] });
};
const ColumnFontStyleAction = ({
  onSelect
}) => {
  const intl = (0, import_react_intl.useIntl)();
  const column = (0, import_DataTableContext.useDataTableColumn)();
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    DataTableActionsMenu.Item,
    {
      onSelect: () => {
        column.toggleFontStyle();
        onSelect?.();
      },
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(DataTableActionsMenu.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.WrapTextIcon, {}) }),
        column.getFontStyle() === "text" ? intl.formatMessage(import_messages.dataTableMessages.fontStyleCodeLabel) : intl.formatMessage(import_messages.dataTableMessages.fontStyleTextLabel)
      ]
    }
  );
};
const DataTableActionsMenu = Object.assign(
  function DataTableActionsMenuInternal(props) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, { children: props.children });
  },
  {
    Intent: import_Intent.Intent,
    Item: import_Item.Item,
    // DEPRECATION: [APPDEV-15604] BREAKING-CHANGE: Remove TableActionsMenu.ItemIcon.
    /**
     * The DataTableActionsMenu.ItemIcon is deprecated, use DataTableActionsMenu.Prefix instead.
     * @deprecated Use `TableActionsMenu.Prefix` instead. This slot will be removed in the future.
     */
    ItemIcon: import_Prefix.Prefix,
    Prefix: import_Prefix.Prefix,
    Suffix: import_Suffix.Suffix,
    Group: import_Group.Group,
    Label: import_Label.Label,
    CopyItem,
    SubMenu: import_Sub.Sub,
    SubContent: import_SubContent.SubContent,
    SubTrigger: import_SubTrigger.SubTrigger,
    Link: import_Link.Link,
    LineWrap: LineWrapAction,
    HideColumn: HideColumnAction,
    ColumnOrder: ColumnOrderAction,
    ColumnPinning: ColumnPinningAction,
    ColumnFontStyle: ColumnFontStyleAction
  }
);
