// @dynatrace/strato-components/tables/DataTable/features/UserActions/useUserActions.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useUserActions_exports = {};
__export(useUserActions_exports, {
  useUserActions: () => useUserActions
});
module.exports = __toCommonJS(useUserActions_exports);
var import_react = require("react");
var import_CellActions = require("./CellActions.js");
var import_ColumnActions = require("./ColumnActions.js");
var import_RowActions = require("./RowActions.js");
function useUserActions(props, options, consumerColumnDefMap) {
  const configuredActions = (0, import_react.useMemo)(() => {
    const actions = {
      cellActions: [],
      columnActions: [],
      rowActions: void 0
    };
    import_react.Children.forEach(props.children, (child) => {
      if ((0, import_CellActions.isDataTableCellActions)(child)) {
        actions.cellActions.push({
          column: child.props.column,
          action: child.props.children
        });
      }
      if ((0, import_ColumnActions.isDataTableColumnActions)(child)) {
        actions.columnActions.push({
          column: child.props.column,
          action: child.props.children
        });
      }
      if ((0, import_RowActions.isDataTableRowActions)(child)) {
        actions.rowActions = child.props.children;
      }
    });
    return actions;
  }, [props.children]);
  options.consumerColumnDefMap = consumerColumnDefMap;
  options.enableColumnActions = configuredActions.columnActions.length > 0;
  options.columnActions = configuredActions.columnActions;
  options.enableCellActions = configuredActions.cellActions.length > 0;
  options.cellActions = configuredActions.cellActions;
  options.enableRowActions = configuredActions.rowActions !== void 0;
  options.rowActions = configuredActions.rowActions;
}
