// @dynatrace/strato-components/filters/filter-field/hooks/useErrorLogsStore.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useErrorLogsStore_exports = {};
__export(useErrorLogsStore_exports, {
  useErrorLogsStore: () => useErrorLogsStore
});
module.exports = __toCommonJS(useErrorLogsStore_exports);
var import_zustand = require("zustand");
var import_middleware = require("zustand/middleware");
const initialState = {
  cursorPosition: 0,
  action: null,
  isControlled: false
};
const useErrorLogsStore = (0, import_zustand.create)(
  (0, import_middleware.combine)(initialState, (set) => ({
    setCursorPosition(pos) {
      set({
        cursorPosition: pos
      });
    },
    setAction(action) {
      set({
        action
      });
    },
    setIsControlled(value) {
      set({
        isControlled: value
      });
    }
  }))
);
