// @dynatrace/strato-components/filters/filter-field/hooks/useFilterFieldLinting.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useFilterFieldLinting_exports = {};
__export(useFilterFieldLinting_exports, {
  useFilterFieldLinting: () => useFilterFieldLinting
});
module.exports = __toCommonJS(useFilterFieldLinting_exports);
var import_lint = require("@codemirror/lint");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_messages = require("../messages.js");
var import_error_message_helpers = require("../utils/error-utils/error-message-helpers.js");
var import_generate_errors_or_diagnostics = require("../utils/generate-errors-or-diagnostics.js");
function useFilterFieldLinting(customTypes) {
  const intl = (0, import_react_intl.useIntl)();
  return (0, import_react.useMemo)(
    () => (0, import_lint.linter)(
      (view) => (0, import_generate_errors_or_diagnostics.generateErrorsOrDiagnostics)(view, {
        emptyGroupError: intl.formatMessage(
          import_messages.filterFieldMessages.linting.emptyGroupError
        ),
        incompleteGroupError: intl.formatMessage(
          import_messages.filterFieldMessages.linting.incompleteGroupError
        ),
        unsupportedExplicitGroup: intl.formatMessage(
          import_messages.filterFieldMessages.linting.unsupportedExplicitGroup
        ),
        unsupportedPositionOperator: intl.formatMessage(
          import_messages.filterFieldMessages.linting.unsupportedPositionOperator
        ),
        invalidWildCardOperatorInList: intl.formatMessage(
          import_messages.filterFieldMessages.linting.invalidWildCardOperatorInList
        ),
        duplicatePositionOperator: intl.formatMessage(
          import_messages.filterFieldMessages.linting.duplicatePositionOperator
        ),
        invalidPositionOperator: intl.formatMessage(
          import_messages.filterFieldMessages.linting.invalidPositionOperator
        ),
        incompleteStatementError: intl.formatMessage(
          import_messages.filterFieldMessages.linting.incompleteStatementError
        ),
        missingKeyError: intl.formatMessage(
          import_messages.filterFieldMessages.linting.missingKeyError
        ),
        missingValueError: intl.formatMessage(
          import_messages.filterFieldMessages.linting.missingValueError
        ),
        missingOperatorError: intl.formatMessage(
          import_messages.filterFieldMessages.linting.missingOperatorError
        ),
        invalidValueErrorGeneric: intl.formatMessage(
          import_messages.filterFieldMessages.linting.invalidValueErrorGeneric
        ),
        invalidValueErrorDurations: intl.formatMessage(
          import_messages.filterFieldMessages.linting.invalidValueErrorDurations
        ),
        invalidValueErrorNumbers: intl.formatMessage(
          import_messages.filterFieldMessages.linting.invalidValueErrorNumbers
        ),
        invalidValueErrorStrings: intl.formatMessage(
          import_messages.filterFieldMessages.linting.invalidValueErrorStrings
        ),
        invalidValueErrorIPAddresses: intl.formatMessage(
          import_messages.filterFieldMessages.linting.invalidValueErrorIPAddresses
        ),
        invalidValueErrorUIDs: intl.formatMessage(
          import_messages.filterFieldMessages.linting.invalidValueErrorUIDs
        ),
        invalidValueErrorTimestamps: intl.formatMessage(
          import_messages.filterFieldMessages.linting.invalidValueErrorTimestamps
        ),
        invalidValueErrorSmartscapeIds: intl.formatMessage(
          import_messages.filterFieldMessages.linting.invalidValueErrorSmartscapeIds
        ),
        invalidValueErrorBoolean: intl.formatMessage(
          import_messages.filterFieldMessages.linting.invalidValueErrorBoolean
        ),
        invalidLogicalOperator: intl.formatMessage(
          import_messages.filterFieldMessages.linting.invalidLogicalOperator
        ),
        unsupportedLogicalOperator: intl.formatMessage(
          import_messages.filterFieldMessages.linting.unsupportedLogicalOperator
        ),
        invalidInclusionValue: intl.formatMessage(
          import_messages.filterFieldMessages.linting.invalidInclusionValue
        ),
        emptyInclusionValue: intl.formatMessage(
          import_messages.filterFieldMessages.linting.emptyInclusionValue
        ),
        missingSpace: intl.formatMessage(
          import_messages.filterFieldMessages.linting.missingSpace
        ),
        invalidKey: intl.formatMessage(
          import_messages.filterFieldMessages.linting.invalidKey
        ),
        invalidWildCardOperator: intl.formatMessage(
          import_messages.filterFieldMessages.linting.invalidWildCardOperator
        ),
        invalidCharacterInIdentifier: intl.formatMessage(
          import_messages.filterFieldMessages.linting.invalidCharacterInIdentifier
        ),
        unsupportedSearchOperator: intl.formatMessage(
          import_messages.filterFieldMessages.linting.unsupportedSearchOperator
        ),
        invalidSearchOperatorInGroup: intl.formatMessage(
          import_messages.filterFieldMessages.linting.invalidSearchOperatorInGroup
        ),
        unsupportedSearchWithLogicalOr: (allowedOperator = "AND") => intl.formatMessage(
          import_messages.filterFieldMessages.linting.unsupportedSearchWithLogicalOr,
          {
            allowedOperator
          }
        ),
        invalidComparatorError: (allowedComparators = "=, !=, >, >=, <, <=") => intl.formatMessage(
          import_messages.filterFieldMessages.linting.invalidComparatorError,
          {
            allowedComparators
          }
        ),
        invalidValueErrorMultipleTypes: (serializedTypes = []) => {
          const typeLabels = serializedTypes.map((type) => {
            const labelKey = (0, import_error_message_helpers.getTypeLabelKey)(type);
            if (labelKey) {
              return intl.formatMessage(
                import_messages.filterFieldMessages.linting.typeLabels[labelKey]
              );
            }
            return (0, import_error_message_helpers.getCustomTypeLabel)(type, customTypes) ?? type;
          });
          const allowedTypes = new Intl.ListFormat(intl.locale, {
            style: "long",
            type: "conjunction"
          }).format(typeLabels);
          return intl.formatMessage(
            import_messages.filterFieldMessages.linting.invalidValueErrorMultipleTypes,
            { allowedTypes }
          );
        },
        incompleteStatementWithEnabledSearchError: (operator = "* ~") => intl.formatMessage(
          import_messages.filterFieldMessages.linting.incompleteStatementWithEnabledSearchError,
          {
            operator
          }
        )
      })
    ),
    [intl, customTypes]
  );
}
