// @dynatrace/strato-components/filters/filter-field/hooks/useSuggestions.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useSuggestions_exports = {};
__export(useSuggestions_exports, {
  stringifySuggestions: () => stringifySuggestions,
  useSuggestions: () => useSuggestions
});
module.exports = __toCommonJS(useSuggestions_exports);
var import_view = require("@codemirror/view");
var import_react = require("react");
var import_react_error_boundary = require("react-error-boundary");
var import_react_intl = require("react-intl");
var import_useErrorLogsStore = require("./useErrorLogsStore.js");
var import_useReferencedState = require("./useReferencedState.js");
var import_usePrevious = require("../../../core/hooks/usePrevious.js");
var import_useFormFieldValidation = require("../../../forms/forms-core/hooks/useFormFieldValidation.js");
var import_editor_config = require("../editor-state-configs/editor-config.js");
var import_messages = require("../messages.js");
var import_get_next_suggestion = require("../utils/get-next-suggestion.js");
var import_get_suggestion_transaction = require("../utils/get-suggestion-transaction.js");
function stringifySuggestions(value) {
  return JSON.stringify(value, replacer);
}
function replacer(_key, value) {
  if ((0, import_react.isValidElement)(value)) {
    const type = value.type;
    return `$$ReactElement:${typeof type === "string" ? type : type.displayName ?? type.name ?? "Component"}`;
  }
  return value;
}
function useSuggestions(props) {
  const {
    ariaDisabled,
    autoSuggestions,
    isOpen,
    id,
    suggestionsOverlayRef,
    editorViewRef,
    onSuggest,
    viewDispatch,
    hiddenInputRef,
    ariaLabelingProps
  } = props;
  const { showBoundary } = (0, import_react_error_boundary.useErrorBoundary)();
  const setUserAction = (0, import_useErrorLogsStore.useErrorLogsStore)((state) => state.setAction);
  const { ariaProps } = (0, import_useFormFieldValidation.useFormFieldValidation)(
    hiddenInputRef,
    {},
    ariaLabelingProps
  );
  const [
    autoSuggestionValues,
    setAutoSuggestionValues,
    autoSuggestionValuesRef
  ] = (0, import_useReferencedState.useReferencedState)({});
  const [
    recentSuggestionValues,
    setRecentSuggestionValues,
    recentSuggestionValuesRef
  ] = (0, import_useReferencedState.useReferencedState)([]);
  const [
    pinnedSuggestionValues,
    setPinnedSuggestionValues,
    pinnedSuggestionValuesRef
  ] = (0, import_useReferencedState.useReferencedState)([]);
  const [
    logicalOperatorSuggestionValues,
    setLogicalOperatorSuggestionValues,
    logicalOperatorSuggestionValuesRef
  ] = (0, import_useReferencedState.useReferencedState)([]);
  const [
    pastedContentSuggestionValues,
    setPastedContentSuggestionValues,
    pastedContentSuggestionValuesRef
  ] = (0, import_useReferencedState.useReferencedState)([]);
  const [
    searchSuggestionValues,
    setSearchSuggestionValues,
    searchSuggestionValuesRef
  ] = (0, import_useReferencedState.useReferencedState)([]);
  const [
    fallbackKeySuggestionValues,
    setFallbackKeySuggestionValues,
    fallbackKeySuggestionValuesRef
  ] = (0, import_useReferencedState.useReferencedState)([]);
  const [
    statementSuggestionValues,
    setStatementSuggestionValues,
    statementSuggestionValuesRef
  ] = (0, import_useReferencedState.useReferencedState)([]);
  const [prevAutoSuggestionValues, setPrevAutoSuggestionValues] = (0, import_react.useState)({});
  const [prevRecentSuggestionValues, setPrevRecentSuggestionValues] = (0, import_react.useState)([]);
  const [
    prevPastedContentSuggestionValues,
    setPrevPastedContentSuggestionValues
  ] = (0, import_react.useState)([]);
  const [prevSearchSuggestionValues, setPrevSearchSuggestionValues] = (0, import_react.useState)([]);
  const [prevFallbackKeySuggestionValues, setPrevFallbackKeySuggestionValues] = (0, import_react.useState)([]);
  const [prevStatementSuggestionValues, setPrevStatementSuggestionValues] = (0, import_react.useState)([]);
  const [focusedListItem, setFocusedListItem] = (0, import_react.useState)();
  const prevFocusedListItem = (0, import_usePrevious.usePrevious)(focusedListItem);
  const suggestionsContext = (0, import_react.useRef)(
    void 0
  );
  const _onSuggest = (0, import_react.useCallback)(
    (args) => {
      suggestionsContext.current = args.suggestionsContext;
      if (autoSuggestions && stringifySuggestions(args.autoSuggestionValues) !== stringifySuggestions(autoSuggestionValuesRef.current)) {
        setAutoSuggestionValues(args.autoSuggestionValues);
      }
      if (stringifySuggestions(args.recentSuggestionValues) !== stringifySuggestions(recentSuggestionValuesRef.current)) {
        setRecentSuggestionValues(args.recentSuggestionValues);
      }
      if (stringifySuggestions(args.pinnedSuggestionValues) !== stringifySuggestions(pinnedSuggestionValuesRef.current)) {
        setPinnedSuggestionValues(args.pinnedSuggestionValues);
      }
      if (stringifySuggestions(args.logicalOperatorSuggestionValues) !== stringifySuggestions(logicalOperatorSuggestionValuesRef.current)) {
        setLogicalOperatorSuggestionValues(
          args.logicalOperatorSuggestionValues
        );
      }
      if (stringifySuggestions(args.pastedContentSuggestionValues) !== stringifySuggestions(pastedContentSuggestionValuesRef.current)) {
        setPastedContentSuggestionValues(args.pastedContentSuggestionValues);
      }
      if (stringifySuggestions(args.searchSuggestionValues) !== stringifySuggestions(searchSuggestionValuesRef.current)) {
        setSearchSuggestionValues(args.searchSuggestionValues);
      }
      if (stringifySuggestions(args.fallbackKeySuggestionValues) !== stringifySuggestions(fallbackKeySuggestionValuesRef.current)) {
        setFallbackKeySuggestionValues(args.fallbackKeySuggestionValues);
      }
      if (stringifySuggestions(args.filterStatementSuggestionValues) !== stringifySuggestions(statementSuggestionValuesRef.current)) {
        setStatementSuggestionValues(args.filterStatementSuggestionValues);
      }
    },
    [
      autoSuggestions,
      autoSuggestionValuesRef,
      recentSuggestionValuesRef,
      pinnedSuggestionValuesRef,
      logicalOperatorSuggestionValuesRef,
      pastedContentSuggestionValuesRef,
      searchSuggestionValuesRef,
      fallbackKeySuggestionValuesRef,
      statementSuggestionValuesRef,
      setAutoSuggestionValues,
      setRecentSuggestionValues,
      setPinnedSuggestionValues,
      setLogicalOperatorSuggestionValues,
      setPastedContentSuggestionValues,
      setSearchSuggestionValues,
      setFallbackKeySuggestionValues,
      setStatementSuggestionValues
    ]
  );
  const [prevOpen, setPrevOpen] = (0, import_react.useState)(isOpen);
  const intl = (0, import_react_intl.useIntl)();
  const contentEditableDivAriaLabel = intl.formatMessage(
    import_messages.filterFieldMessages.ariaLabel
  );
  const suggestionsOverlayHintId = "suggestions-overlay-announcement";
  const ariaDescribedBys = ariaProps["aria-describedby"] ?? "";
  const suggestionsHintId = isOpen ? suggestionsOverlayHintId : "";
  const ariaLabelProp = ariaLabelingProps?.["aria-label"];
  const ariaDetailsProp = ariaLabelingProps?.["aria-details"] ?? "";
  const ariaLabelledbyProp = ariaLabelingProps?.["aria-labelledby"] ?? "";
  const ariaAttributes = (0, import_react.useMemo)(
    () => ({
      "aria-label": ariaLabelProp ?? contentEditableDivAriaLabel,
      "aria-autocomplete": "list",
      "aria-haspopup": "listbox",
      ...isOpen && {
        "aria-controls": `suggestions-${id}`
      },
      ...focusedListItem?.id && {
        "aria-activedescendant": focusedListItem?.id
      },
      ...ariaDisabled && {
        "aria-disabled": `${ariaDisabled}`
      },
      "aria-describedby": (ariaDescribedBys + " " + suggestionsHintId).trim(),
      ...ariaDetailsProp && {
        "aria-details": `${ariaDetailsProp}`
      },
      ...ariaLabelledbyProp && {
        "aria-labelledby": `${ariaLabelledbyProp}`
      }
    }),
    [
      ariaDescribedBys,
      ariaDisabled,
      ariaLabelProp,
      ariaDetailsProp,
      ariaLabelledbyProp,
      contentEditableDivAriaLabel,
      focusedListItem,
      isOpen,
      suggestionsHintId,
      id
    ]
  );
  const applySuggestion = (0, import_react.useCallback)(
    (element, value, trailing, insertionStrategy, cursorBeforeTrailing, range, shouldEscapeSuggestion) => {
      if (!editorViewRef.current || !suggestionsContext.current || !element) {
        return;
      }
      setUserAction({
        type: "applySuggestion",
        data: {
          value,
          trailing,
          insertionStrategy,
          cursorBeforeTrailing
        }
      });
      const suggestionTransaction = (0, import_get_suggestion_transaction.getSuggestionTransaction)(
        value,
        range,
        shouldEscapeSuggestion,
        trailing,
        insertionStrategy,
        cursorBeforeTrailing,
        suggestionsContext.current,
        editorViewRef.current.state
      );
      viewDispatch({
        ...suggestionTransaction,
        annotations: import_editor_config.stringAnnotation.of(import_editor_config.ANNOTATION_SUGGESTION_APPLY)
      });
      setFocusedListItem(void 0);
    },
    // Excluding suggestionsContext. We are accessing the current context by reference, will always be up to date.
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [setFocusedListItem, viewDispatch]
  );
  if (prevOpen && !isOpen) {
    queueMicrotask(() => {
      try {
        onSuggest?.();
      } catch (e) {
        showBoundary(e);
      }
    });
  }
  if (prevOpen !== isOpen || autoSuggestionValuesRef.current !== prevAutoSuggestionValues || recentSuggestionValuesRef.current !== prevRecentSuggestionValues || pastedContentSuggestionValuesRef.current !== prevPastedContentSuggestionValues || searchSuggestionValuesRef.current !== prevSearchSuggestionValues || fallbackKeySuggestionValuesRef.current !== prevFallbackKeySuggestionValues || statementSuggestionValuesRef.current !== prevStatementSuggestionValues) {
    if (isOpen) {
      if (!(0, import_get_next_suggestion.getCurrentFocusableListItem)(
        suggestionsOverlayRef.current,
        focusedListItem?.id
      )) {
        setFocusedListItem(void 0);
      }
    }
    setPrevAutoSuggestionValues(autoSuggestionValuesRef.current ?? {});
    setPrevRecentSuggestionValues(recentSuggestionValuesRef.current ?? []);
    setPrevPastedContentSuggestionValues(
      pastedContentSuggestionValuesRef.current ?? []
    );
    setPrevSearchSuggestionValues(searchSuggestionValuesRef.current ?? []);
    setPrevFallbackKeySuggestionValues(
      fallbackKeySuggestionValuesRef.current ?? []
    );
    setPrevStatementSuggestionValues(
      statementSuggestionValuesRef.current ?? []
    );
    setPrevOpen(isOpen);
  }
  if (prevFocusedListItem !== focusedListItem || prevOpen !== isOpen) {
    viewDispatch({
      effects: import_editor_config.ariaAttributesCompartment.reconfigure(
        import_view.EditorView.contentAttributes.of(ariaAttributes)
      )
    });
  }
  const handleNextSuggestion = (nextSuggestion) => {
    if (nextSuggestion) {
      const id2 = nextSuggestion.dataset.stratoFilterFieldVirtualFocusId;
      const type = nextSuggestion.dataset.stratoFilterFieldVirtualFocusType;
      setFocusedListItem({ id: id2, type });
      nextSuggestion.scrollIntoView({ block: "center" });
    }
  };
  const autoSuggestionsConfig = (0, import_react.useMemo)(() => {
    const filterValue = suggestionsContext.current?.currentStatement?.value;
    const cursorPosition = editorViewRef.current?.state.selection.main.head;
    return {
      trailing: cursorPosition === filterValue?.range.to ? " " : void 0,
      groupedSuggestions: autoSuggestionValues
    };
  }, [autoSuggestionValues]);
  return {
    ariaAttributes,
    autoSuggestionsConfig,
    recentSuggestionValues,
    pinnedSuggestionValues,
    logicalOperatorSuggestionValues,
    pastedContentSuggestionValues,
    searchSuggestionValues,
    fallbackKeySuggestionValues,
    statementSuggestionValues,
    focusedListItem,
    suggestionsContext,
    applySuggestion,
    handleNextSuggestion,
    _onSuggest,
    suggestionsOverlayHintId
  };
}
