// @dynatrace/strato-components/filters/filter-field/hooks/useShowMore.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useShowMore_exports = {};
__export(useShowMore_exports, {
  sliceFlattenedSuggestions: () => sliceFlattenedSuggestions,
  useShowMore: () => useShowMore
});
module.exports = __toCommonJS(useShowMore_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_useId = require("../../../core/hooks/useId.js");
var import_getVirtualFocusStyles = require("../../../core/styles/getVirtualFocusStyles.js");
var import_Flex = require("../../../layouts/flex/Flex.js");
var import_messages = require("../messages.js");
var import_useFilterFieldContext = require("./useFilterFieldContext.js");
var import_get_next_suggestion = require("../utils/get-next-suggestion.js");
function useShowMore(props) {
  const { items, primaryCount, sliceFn: _propSliceFn } = props;
  const sliceFn = _propSliceFn ?? defaultSlice;
  const [isExpanded, setIsExpanded] = (0, import_react.useState)(false);
  const [pendingNextSuggestionIndex, setPendingNextSuggestionIndex] = (0, import_react.useState)(null);
  const { focusedListItem, handleNextSuggestion, suggestionsOverlayRef } = (0, import_useFilterFieldContext.useFilterFieldContext)();
  const id = (0, import_useId.useId)();
  (0, import_react.useLayoutEffect)(() => {
    if (pendingNextSuggestionIndex !== null) {
      const element = (0, import_get_next_suggestion.getCurrentFocusableListItemByIndex)(
        suggestionsOverlayRef,
        pendingNextSuggestionIndex
      );
      handleNextSuggestion(element);
      setPendingNextSuggestionIndex(null);
    }
  }, [pendingNextSuggestionIndex, suggestionsOverlayRef, handleNextSuggestion]);
  const controlProps = (0, import_react.useMemo)(
    () => ({
      // Role needs to be option, because otherwise NVDA will crash if the activeDecendant of a listbox is anything other than a role option or group.
      role: "option",
      onClick: (event) => {
        event.preventDefault();
        event.stopPropagation();
        const isKeyboard = event.detail === 0 || event.nativeEvent instanceof KeyboardEvent;
        if (isExpanded || !isKeyboard) {
          setIsExpanded((previousState) => !previousState);
          return;
        }
        const index = (0, import_get_next_suggestion.getCurrentFocusableListItemIndex)(
          suggestionsOverlayRef,
          focusedListItem?.id
        );
        setIsExpanded((previousState) => !previousState);
        setPendingNextSuggestionIndex(index ?? null);
      },
      id,
      "data-strato-filter-field-virtual-focus-id": id,
      "data-strato-filter-field-virtual-focus-type": "button",
      "data-dt-name": "ShowMoreLess",
      "data-dt-properties": `state:${isExpanded ? "less" : "more"}`,
      className: (0, import_getVirtualFocusStyles.getVirtualFocusStyles)(id === focusedListItem?.id).className,
      width: "full",
      textAlign: "start",
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { gap: 2, as: "span", paddingLeft: 4, children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_react_intl.FormattedMessage,
          {
            ...import_messages.filterFieldMessages.suggestions.showMore,
            values: { isExpanded }
          }
        ),
        isExpanded ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ChevronUpSmallIcon, {}) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ChevronDownSmallIcon, {})
      ] })
    }),
    [focusedListItem?.id, id, isExpanded, suggestionsOverlayRef]
  );
  const { items: resultItems, isSliced } = (0, import_react.useMemo)(
    () => sliceFn(items, primaryCount),
    // Only recompute when the suggestions themselves change to avoid flashing
    // when the primaryCount toggles (e.g., empty vs filled input) without data changes.
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [items, sliceFn]
  );
  if (primaryCount === -1) {
    return {
      items,
      isSliced: false,
      isExpanded,
      controlProps
    };
  }
  return {
    items: isExpanded ? items : resultItems,
    isSliced,
    controlProps,
    isExpanded
  };
}
function defaultSlice(items, count) {
  if (items.length <= count + 1) {
    return { items, isSliced: false };
  }
  return { items: items.slice(0, count), isSliced: true };
}
function sliceFlattenedSuggestions(suggestions, count) {
  if (count === -1) {
    return { items: suggestions, isSliced: false };
  }
  let totalSuggestionCount = 0;
  for (const item of suggestions) {
    if (item.kind !== "group-header") {
      totalSuggestionCount++;
    }
  }
  if (totalSuggestionCount <= count + 1) {
    return { items: suggestions, isSliced: false };
  }
  let visibleCount = 0;
  const itemsShown = [];
  for (const item of suggestions) {
    if (item.kind === "group-header") {
      itemsShown.push(item);
      continue;
    }
    if (visibleCount >= count) {
      break;
    }
    itemsShown.push(item);
    visibleCount++;
  }
  if (itemsShown.length > 0 && itemsShown[itemsShown.length - 1].kind === "group-header") {
    itemsShown.pop();
  }
  return { items: itemsShown, isSliced: true };
}
