// @dynatrace/strato-components/core/time/timeframeStore/actions/calendar.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var calendar_exports = {};
__export(calendar_exports, {
  getCalendarActions: () => getCalendarActions,
  getTimeframeFromCalendarSelection: () => getTimeframeFromCalendarSelection
});
module.exports = __toCommonJS(calendar_exports);
var import_date_fns = require("date-fns");
var import_calendarMessages = require("../../../../forms/calendar/calendarMessages.js");
var import_helpers = require("../../../../forms/forms-core/time/helpers.js");
var import_helpers2 = require("../../helpers.js");
var import_validation = require("../../validation.js");
function getTimeframeFromCalendarSelection(inputValue, state) {
  if (!(0, import_helpers2.isTimeframe)(state.calendarValue, "range")) {
    return null;
  }
  if (inputValue === null) {
    return null;
  }
  const stringDate = inputValue.toISOString();
  const calendarRangeValue = state.calendarValue;
  const { from, to } = calendarRangeValue ?? {};
  let position = "from";
  let newTimeframe = {
    from: null,
    to: null
  };
  if (from && !to) {
    const isToBeforeFrom = (0, import_date_fns.isBefore)(stringDate, from.absoluteDate);
    if (!isToBeforeFrom) {
      newTimeframe = {
        from,
        to: null
      };
      position = "to";
    }
    if (isToBeforeFrom) {
      newTimeframe = {
        from: null,
        to: (0, import_helpers.getTimeValueFromSimpleValue)(
          from.absoluteDate,
          state.timeOffset,
          "to",
          void 0,
          state.min,
          state.max,
          state.precision
        )
      };
    }
  }
  const newTimeValue = (0, import_helpers.getTimeValueFromSimpleValue)(
    stringDate,
    state.timeOffset,
    position,
    void 0,
    state.min,
    state.max,
    state.precision
  );
  newTimeframe[position] = newTimeValue;
  return newTimeframe;
}
function getCalendarActions(isControlled, type, formFieldMessagesDispatch, onChangeProp) {
  return (set) => ({
    setCalendarValue: (inputValue) => {
      set((state) => {
        if (type === "range") {
          const newTimeframe = getTimeframeFromCalendarSelection(
            inputValue,
            state
          );
          if ((0, import_helpers2.isFullTimeframe)(newTimeframe) || newTimeframe === null) {
            onChangeProp?.(newTimeframe);
            if (isControlled) {
              return state;
            }
          }
          const calendarError2 = (0, import_validation.validate)(newTimeframe, state.min, state.max, {
            formMessageDispatch: formFieldMessagesDispatch,
            formerError: state.calendarError,
            precision: state.precision,
            showError: false,
            checkOrder: true,
            ref: state.calendarRef,
            valueMode: "timeframeSelector",
            errorMessages: import_calendarMessages.CALENDAR_ERROR_MESSAGES
          });
          return {
            calendarValue: newTimeframe,
            calendarError: calendarError2
          };
        }
        const newTimeValue = (0, import_helpers.getTimeValueFromSimpleValue)(
          inputValue,
          state.timeOffset,
          void 0,
          void 0,
          state.min,
          state.max,
          state.precision
        ) ?? null;
        onChangeProp?.(
          newTimeValue?.value ?? null
        );
        if (isControlled) {
          return state;
        }
        const calendarError = (0, import_validation.validate)(newTimeValue, state.min, state.max, {
          formMessageDispatch: formFieldMessagesDispatch,
          formerError: state.calendarError,
          precision: state.precision,
          showError: false,
          checkOrder: true,
          ref: state.calendarRef,
          valueMode: "single",
          errorMessages: import_calendarMessages.CALENDAR_ERROR_MESSAGES
        });
        return { calendarValue: newTimeValue, calendarError };
      });
    }
  });
}
