// @dynatrace/strato-components/core/time/timeframeStore/actions/dateTimePicker.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var dateTimePicker_exports = {};
__export(dateTimePicker_exports, {
  getDateTimePickerActions: () => getDateTimePickerActions
});
module.exports = __toCommonJS(dateTimePicker_exports);
var import_focusDateTimeInput = require("../../../../forms/date-time-picker/utils/focusDateTimeInput.js");
var import_helpers = require("../../../../forms/forms-core/time/helpers.js");
var import_timeParser = require("../../../../forms/forms-core/time/timeParser.js");
var import_validation = require("../../validation.js");
function getDateTimePickerActions(formFieldMessagesDispatch, onChangeProp, isControlled) {
  const actions = (set) => {
    return {
      setCalendarValue: (inputValue) => {
        set((state) => {
          const newTimeValue = (0, import_helpers.getTimeValueFromSimpleValue)(
            inputValue,
            state.timeOffset,
            void 0,
            void 0,
            state.min,
            state.max,
            state.precision
          ) ?? null;
          const calendarError = (0, import_validation.validate)(newTimeValue, state.min, state.max, {
            formMessageDispatch: formFieldMessagesDispatch,
            formerError: state.dateTimePickerError,
            precision: state.precision,
            showError: false,
            checkOrder: true,
            ref: state.dateTimePickerRef,
            expectedFormat: state.formatString,
            valueMode: "single",
            required: state.required
          });
          state.dateTimePickerRef?.current?.close();
          (0, import_focusDateTimeInput.focusElementWithRetry)(
            () => state.dateTimeInputRef.current?.spinButtonHandlers.get("hour")?.element.current,
            state.dateTimeInputRef
          );
          const calendarTimeValue = (0, import_helpers.toCalendarValue)(
            newTimeValue,
            state.precision
          );
          return {
            calendarValue: calendarTimeValue,
            dateTimePickerValue: newTimeValue,
            dateTimePickerError: calendarError,
            calendarError
          };
        });
      },
      setDateTimePickerValue: (value) => {
        return set((state) => {
          const newValue = (0, import_timeParser.getTimeValue)(value, state.precision) ?? null;
          const isPartialValue = newValue?.absoluteDate === "";
          if (isPartialValue) {
            const errorKey = (0, import_validation.validate)(newValue, state.min, state.max, {
              formMessageDispatch: formFieldMessagesDispatch,
              formerError: state.dateTimePickerError,
              precision: state.precision,
              showError: false,
              checkOrder: true,
              ref: state.dateTimePickerRef,
              expectedFormat: state.formatString,
              valueMode: "single"
            });
            onChangeProp?.("");
            return {
              calendarError: errorKey,
              dateTimePickerError: errorKey
            };
          }
          onChangeProp?.(newValue);
          const calendarTimeValue = (0, import_helpers.toCalendarValue)(newValue, state.precision);
          if (isControlled) {
            return { calendarValue: calendarTimeValue };
          }
          const dateTimePickerError = (0, import_validation.validate)(newValue, state.min, state.max, {
            formMessageDispatch: formFieldMessagesDispatch,
            formerError: state.dateTimePickerError,
            precision: state.precision,
            showError: false,
            checkOrder: false,
            ref: state.dateTimePickerRef,
            expectedFormat: state.formatString,
            valueMode: "single"
          });
          return {
            dateTimePickerValue: newValue,
            calendarValue: calendarTimeValue,
            calendarError: dateTimePickerError,
            dateTimePickerError
          };
        });
      }
    };
  };
  return actions;
}
