// @dynatrace/strato-components/core/time/timeframeStore/actions/dateTimeRangePicker.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var dateTimeRangePicker_exports = {};
__export(dateTimeRangePicker_exports, {
  getDateTimeRangePickerActions: () => getDateTimeRangePickerActions
});
module.exports = __toCommonJS(dateTimeRangePicker_exports);
var import_calendar = require("./calendar.js");
var import_focusDateTimeInput = require("../../../../forms/date-time-picker/utils/focusDateTimeInput.js");
var import_timeParser = require("../../../../forms/forms-core/time/timeParser.js");
var import_helpers = require("../../helpers.js");
var import_validation = require("../../validation.js");
function getDateTimeRangePickerActions(formFieldMessagesDispatch, onChangeProp) {
  const actions = (set) => {
    return {
      setCalendarValue: (inputValue) => {
        set((state) => {
          const calendarTimeframe = (0, import_calendar.getTimeframeFromCalendarSelection)(
            inputValue,
            state
          );
          if ((0, import_helpers.isFullTimeframe)(calendarTimeframe)) {
            state.dateTimeRangePickerRef?.current?.close();
            (0, import_focusDateTimeInput.focusElementWithRetry)(
              () => state.dateTimeFromRef.current?.spinButtonHandlers.get("hour")?.element.current,
              state.dateTimeFromRef
            );
            onChangeProp?.(calendarTimeframe);
            return {
              calendarValue: calendarTimeframe,
              dateTimeRangePickerTimeframe: calendarTimeframe
            };
          }
          return { calendarValue: calendarTimeframe };
        });
      },
      setDateTimeRangePickerTimeframe: (value) => {
        return set((state) => {
          if (value === null) {
            onChangeProp?.(null);
            return {
              dateTimeRangePickerTimeframe: null,
              calendarValue: null
            };
          }
          const merged = (0, import_helpers.mergeRangeValue)(
            state.dateTimeRangePickerTimeframe,
            value
          );
          const timeframeWithPrecision = (0, import_timeParser.removePrecisionFromTimeFrame)(merged, state.precision) ?? null;
          if (timeframeWithPrecision?.from === null && timeframeWithPrecision?.to === null) {
            onChangeProp?.(null);
          }
          (0, import_helpers.isFullTimeframe)(timeframeWithPrecision) && onChangeProp?.(timeframeWithPrecision);
          const isPartialFrom = (0, import_helpers.isTimeValue)(merged.from) && merged.from.absoluteDate === "";
          const isPartialTo = (0, import_helpers.isTimeValue)(merged.to) && merged.to.absoluteDate === "";
          if (isPartialFrom || isPartialTo) {
            const errorKey = (0, import_validation.validate)(
              timeframeWithPrecision,
              state.min,
              state.max,
              {
                formMessageDispatch: formFieldMessagesDispatch,
                formerError: state.dateTimeRangePickerError,
                precision: state.precision,
                showError: false,
                checkOrder: true,
                ref: state.dateTimeRangePickerRef,
                expectedFormat: state.formatString,
                valueMode: "dateTimePicker",
                required: state.required
              }
            );
            const calendarValue = {
              from: isPartialFrom ? null : timeframeWithPrecision?.from ?? null,
              to: isPartialTo ? null : timeframeWithPrecision?.to ?? null
            };
            return {
              dateTimeRangePickerTimeframe: timeframeWithPrecision,
              calendarValue,
              calendarError: errorKey,
              dateTimeRangePickerError: errorKey
            };
          }
          const dateTimeRangePickerError = (0, import_validation.validate)(
            timeframeWithPrecision,
            state.min,
            state.max,
            {
              formMessageDispatch: formFieldMessagesDispatch,
              formerError: state.dateTimeRangePickerError,
              precision: state.precision,
              showError: false,
              checkOrder: true,
              ref: state.dateTimeRangePickerRef,
              expectedFormat: state.formatString,
              valueMode: "dateTimePicker",
              required: state.required
            }
          );
          return {
            dateTimeRangePickerTimeframe: timeframeWithPrecision,
            calendarValue: timeframeWithPrecision,
            calendarError: dateTimeRangePickerError,
            dateTimeRangePickerError
          };
        });
      },
      clearCalendarValue: () => {
        set((state) => {
          if (!state.calendarValue || state.calendarValue.from && state.calendarValue.to) {
            return state;
          }
          if (state.dateTimeRangePickerTimeframe?.from && state.dateTimeRangePickerTimeframe?.to) {
            return { calendarValue: state.dateTimeRangePickerTimeframe };
          }
          return { calendarValue: null };
        });
      }
    };
  };
  return actions;
}
