// @dynatrace/strato-components/core/time/timeframeStore/actions/timeframeSelector.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var timeframeSelector_exports = {};
__export(timeframeSelector_exports, {
  getTimeframeSelectorActions: () => getTimeframeSelectorActions
});
module.exports = __toCommonJS(timeframeSelector_exports);
var import_tz = require("@date-fns/tz");
var import_date_fns = require("date-fns");
var import_user_preferences = require("@dynatrace-sdk/user-preferences");
var import_calendar = require("./calendar.js");
var import_focusDateTimeInput = require("../../../../forms/date-time-picker/utils/focusDateTimeInput.js");
var import_timeParser = require("../../../../forms/forms-core/time/timeParser.js");
var import_helpers = require("../../helpers.js");
var import_validation = require("../../validation.js");
function handleTimeframeSelectorTimeframeChange(inputValue, state, isControlled, timeframeSelectorRef, formFieldMessagesDispatch, onChangeProp, errorMessages) {
  const timeframeWithPrecision = (0, import_timeParser.removePrecisionFromTimeFrame)(inputValue, state.precision) ?? null;
  const value = (0, import_helpers.isFullTimeframe)(timeframeWithPrecision) ? timeframeWithPrecision : null;
  onChangeProp?.(value);
  timeframeSelectorRef.current?.close();
  if (isControlled) {
    return state;
  }
  const error = (0, import_validation.validate)(timeframeWithPrecision, state.min, state.max, {
    formMessageDispatch: formFieldMessagesDispatch,
    formerError: state.timeframeSelectorError,
    precision: state.precision,
    showError: false,
    checkOrder: true,
    ref: timeframeSelectorRef,
    expectedFormat: state.formatString,
    valueMode: "timeframeSelector",
    required: state.required,
    errorMessages
  });
  return {
    timeframeSelectorTimeframe: value,
    calendarValue: value ?? timeframeWithPrecision,
    dateTimeRangePickerTimeframe: value ?? timeframeWithPrecision,
    timeframeSelectorError: error,
    calendarError: error,
    dateTimeRangePickerError: error
  };
}
function handlePreviousTimeframeChange(value) {
  return {
    referenceTimeframe: value
  };
}
function isFullDayTimeframe(timeframe, state) {
  const timeZone = (0, import_user_preferences.getTimezone)();
  const to = new import_tz.TZDateMini(timeframe.to.absoluteDate, timeZone);
  const from = new import_tz.TZDateMini(timeframe.from.absoluteDate, timeZone);
  const isToAtEndOfDay = (0, import_timeParser.removePrecision)((0, import_date_fns.endOfDay)(to).toISOString(), state.precision) === (0, import_timeParser.removePrecision)(timeframe.to.absoluteDate, state.precision);
  const isToAtMax = state.max?.absoluteDate === to.toISOString();
  const isFromAtStartOfDay = (0, import_timeParser.removePrecision)((0, import_date_fns.startOfDay)(from).toISOString(), state.precision) === (0, import_timeParser.removePrecision)(timeframe.from.absoluteDate, state.precision);
  const isFromAtMin = state.min?.absoluteDate === from.toISOString();
  return Boolean(
    (isToAtEndOfDay || isToAtMax) && (isFromAtStartOfDay || isFromAtMin)
  );
}
function getTimeframeSelectorActions(isControlled, timeframeSelectorRef, formFieldMessagesDispatch, onChangeProp, debouncedOnChangeRef, errorMessages) {
  const actions = (set) => {
    return {
      setCalendarValue: (inputValue) => {
        set((state) => {
          const calendarValue = (0, import_calendar.getTimeframeFromCalendarSelection)(
            inputValue,
            state
          );
          if ((0, import_helpers.isFullTimeframe)(calendarValue)) {
            state.dateTimeRangePickerRef?.current?.close();
            const isFullDay = isFullDayTimeframe(calendarValue, state);
            (0, import_focusDateTimeInput.focusElementWithRetry)(
              () => state.dateTimeFromRef.current?.spinButtonHandlers.get("hour")?.element.current,
              state.dateTimeFromRef
            );
            return {
              calendarValue,
              dateTimeRangePickerTimeframe: calendarValue,
              isFullDay
            };
          }
          return { calendarValue, isFullDay: false };
        });
      },
      setDateTimeRangePickerTimeframe: (value) => {
        return set((state) => {
          if (value === null) {
            return {
              dateTimeRangePickerTimeframe: null,
              calendarValue: null,
              dateTimeRangePickerError: void 0,
              calendarError: void 0
            };
          }
          const merged = (0, import_helpers.mergeRangeValue)(
            state.dateTimeRangePickerTimeframe,
            value
          );
          const timeframeWithPrecision = (0, import_timeParser.removePrecisionFromTimeFrame)(merged, state.precision) ?? null;
          const isPartialFrom = (0, import_helpers.isTimeValue)(merged.from) && merged.from.absoluteDate === "";
          const isPartialTo = (0, import_helpers.isTimeValue)(merged.to) && merged.to.absoluteDate === "";
          if (isPartialFrom || isPartialTo) {
            const errorKey = (0, import_validation.validate)(
              timeframeWithPrecision,
              state.min,
              state.max,
              {
                formerError: state.dateTimeRangePickerError,
                precision: state.precision,
                showError: false,
                checkOrder: true,
                ref: state.dateTimeRangePickerRef,
                expectedFormat: state.formatString,
                valueMode: "timeframeSelector",
                required: state.required,
                errorMessages
              }
            );
            const calendarValue = {
              from: isPartialFrom ? null : timeframeWithPrecision?.from ?? null,
              to: isPartialTo ? null : timeframeWithPrecision?.to ?? null
            };
            return {
              dateTimeRangePickerTimeframe: timeframeWithPrecision,
              calendarValue,
              calendarError: errorKey,
              dateTimeRangePickerError: errorKey
            };
          }
          const dateTimeRangePickerError = (0, import_validation.validate)(
            timeframeWithPrecision,
            state.min,
            state.max,
            {
              formerError: state.dateTimeRangePickerError,
              precision: state.precision,
              showError: false,
              checkOrder: true,
              ref: state.dateTimeRangePickerRef,
              expectedFormat: state.formatString,
              valueMode: "timeframeSelector",
              required: state.required,
              errorMessages
            }
          );
          return {
            dateTimeRangePickerTimeframe: timeframeWithPrecision,
            calendarValue: timeframeWithPrecision,
            calendarError: dateTimeRangePickerError,
            dateTimeRangePickerError
          };
        });
      },
      setTimeframeSelectorTimeframe: (inputValue) => {
        debouncedOnChangeRef?.current?.cancel();
        return set(
          (state) => handleTimeframeSelectorTimeframeChange(
            inputValue,
            state,
            isControlled,
            timeframeSelectorRef,
            formFieldMessagesDispatch,
            onChangeProp,
            errorMessages
          )
        );
      },
      setReferenceTimeframe: (value) => {
        set(() => handlePreviousTimeframeChange(value));
      },
      setStepperValue: (value) => {
        set(() => {
          debouncedOnChangeRef?.current?.(value);
          return {
            timeframeSelectorTimeframe: value,
            calendarValue: value,
            dateTimeRangePickerTimeframe: value
          };
        });
      },
      clearCalendarValue: () => {
        set((state) => {
          if (!state.calendarValue || state.calendarValue.from && state.calendarValue.to) {
            return state;
          }
          if (state.dateTimeRangePickerTimeframe?.from && state.dateTimeRangePickerTimeframe?.to) {
            return { calendarValue: state.dateTimeRangePickerTimeframe };
          }
          return { calendarValue: null };
        });
      }
    };
  };
  return actions;
}
