// @dynatrace/strato-components/charts/xy-chart/components/annotations/XYChartAnnotationsArea.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var XYChartAnnotationsArea_exports = {};
__export(XYChartAnnotationsArea_exports, {
  XYChartAnnotationsArea: () => XYChartAnnotationsArea
});
module.exports = __toCommonJS(XYChartAnnotationsArea_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_useBehavioralTrackingProps = require("../../../../core/hooks/useBehavioralTrackingProps.js");
var import_AnnotationsNumericalDisplay = require("../../../core/components/annotations/components/displays/AnnotationsNumericalDisplay.js");
var import_AnnotationsTimeDisplay = require("../../../core/components/annotations/components/displays/AnnotationsTimeDisplay.js");
var import_AnnotationsScrollWrapper = require("../../../core/components/annotations/components/overlays/AnnotationsScrollWrapper.js");
var import_constants = require("../../../core/components/annotations/constants.js");
var import_ChartGridRoot = require("../../../core/components/chart-grid/ChartGridRoot.js");
var import_useGridPlotSize = require("../../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_defaults = require("../../constants/defaults.js");
var import_useXYChartAnnotationsConfig = require("../../hooks/useXYChartAnnotationsConfig.js");
var import_useXYChartAnnotationsData = require("../../hooks/useXYChartAnnotationsData.js");
var import_useXYChartData = require("../../hooks/useXYChartData.js");
var import_useXYChartScales = require("../../hooks/useXYChartScales.js");
var import_selectors = require("../../store/selectors.js");
var import_guards = require("../../types/guards.js");
var import_get_scale = require("../../utils/get-scale.js");
var import_grid_templates = require("../../utils/grid-templates.js");
const XYChartAnnotationsArea = () => {
  const scales = (0, import_useXYChartScales.useXYChartScales)();
  const annotationsConfig = (0, import_useXYChartAnnotationsConfig.useXYChartAnnotationsConfig)();
  const annotationsData = (0, import_useXYChartAnnotationsData.useXYChartAnnotationsData)();
  const setState = (0, import_selectors.useSetState)();
  const { width } = (0, import_useGridPlotSize.useGridPlotSize)();
  const { axes } = (0, import_useXYChartData.useXYChartData)();
  const [behaviorTrackingProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(
    annotationsConfig ?? {}
  );
  const handleUpdateIndicators = (event) => {
    setState((prev) => ({
      ...prev,
      annotations: {
        ...prev.annotations,
        ...event
      }
    }));
  };
  if ((0, import_lodash_es.isUndefined)(annotationsConfig) || (0, import_lodash_es.isUndefined)(annotationsData)) {
    return null;
  }
  const { xAxisId, showLabels = false, annotationsActions } = annotationsConfig;
  const axis = axes.find(({ id }) => id === xAxisId);
  if (axis === void 0) {
    return null;
  }
  const scale = (0, import_get_scale.getCurrentScale)(scales, xAxisId);
  const trackNames = Array.from(annotationsData.keys());
  const serializedAnnotationsData = trackNames.map((name) => annotationsData.get(name)).filter((track) => !(0, import_lodash_es.isUndefined)(track));
  const DisplayComponent = (0, import_guards.isTimeScale)({
    axisType: axis.type,
    axisScale: scale
  }) ? import_AnnotationsTimeDisplay.AnnotationsTimeDisplay : import_AnnotationsNumericalDisplay.AnnotationsNumericalDisplay;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ChartGridRoot.ChartGrid.GridArea,
    {
      paddingBottom: import_defaults.ANNOTATIONS_PADDING_BOTTOM,
      gridArea: import_grid_templates.AREAS.annotations,
      children: ({ availableWidth, availableHeight }) => {
        const labelsOffset = availableWidth - width;
        const annotationsCommonConfig = {
          tracks: serializedAnnotationsData,
          showLabels,
          labelsOffset,
          width: availableWidth,
          markerGroupHover: handleUpdateIndicators,
          textOverflow: import_constants.DEFAULT_INTEGRATION_TEXT_OVERFLOW,
          truncateMode: "middle",
          tooltipConfig: annotationsConfig.tooltip,
          ...behaviorTrackingProps
        };
        return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_AnnotationsScrollWrapper.AnnotationsScrollWrapper,
          {
            height: availableHeight,
            width: availableWidth + import_constants.SCROLLBAR_THRESHOLD,
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              DisplayComponent,
              {
                scale,
                annotationsActions: (annotation) => annotation && annotationsActions?.(annotation),
                ...annotationsCommonConfig
              }
            )
          }
        );
      }
    }
  );
};
