// @dynatrace/strato-components/charts/xy-chart/components/annotations/XYChartAnnotationsIndicators.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var XYChartAnnotationsIndicators_exports = {};
__export(XYChartAnnotationsIndicators_exports, {
  XYChartAnnotationsIndicators: () => XYChartAnnotationsIndicators
});
module.exports = __toCommonJS(XYChartAnnotationsIndicators_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_IndicatorsRenderer = require("../../../core/components/annotations/components/indicators/IndicatorsRenderer.js");
var import_get_unique_indicators_by_priority = require("../../../core/components/annotations/components/indicators/utils/get-unique-indicators-by-priority.js");
var import_get_min_max_as_number = require("../../../core/components/annotations/utils/get-min-max-as-number.js");
var import_ChartGrid_css = require("../../../core/components/chart-grid/ChartGrid.sty.js");
var import_useXYChartAnnotationsConfig = require("../../hooks/useXYChartAnnotationsConfig.js");
var import_useXYChartAnnotationsShownIndicators = require("../../hooks/useXYChartAnnotationsShownIndicators.js");
var import_useXYChartData = require("../../hooks/useXYChartData.js");
var import_useXYChartScales = require("../../hooks/useXYChartScales.js");
var import_guards = require("../../types/guards.js");
var import_get_scale = require("../../utils/get-scale.js");
const XYChartAnnotationsIndicators = () => {
  const indicators = (0, import_useXYChartAnnotationsShownIndicators.useXYChartAnnotationsShownIndicators)();
  const annotationsConfig = (0, import_useXYChartAnnotationsConfig.useXYChartAnnotationsConfig)();
  const scales = (0, import_useXYChartScales.useXYChartScales)();
  const { axes } = (0, import_useXYChartData.useXYChartData)();
  const axisConfig = axes.find(
    (axis) => axis.id === annotationsConfig?.xAxisId
  );
  if ((0, import_lodash_es.isUndefined)(annotationsConfig) || (0, import_lodash_es.isUndefined)(axisConfig)) {
    return;
  }
  const xScale = (0, import_get_scale.getCurrentScale)(scales, axisConfig.id);
  const annotationsScale = {
    axisType: axisConfig.type,
    axisScale: xScale
  };
  if (!(0, import_guards.isAnyNumericalScale)(annotationsScale)) {
    return;
  }
  const [domainStart, domainEnd] = annotationsScale.axisScale.domain();
  const isInverted = domainStart > domainEnd;
  const { min: domainMin, max: domainMax } = (0, import_get_min_max_as_number.getMinMaxAsNumber)(
    domainStart,
    domainEnd
  );
  const uniqueIndicators = (0, import_get_unique_indicators_by_priority.getUniqueIndicatorsByPriority)(
    [indicators],
    [domainMin, domainMax]
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("svg", { className: import_ChartGrid_css.plotLayout, children: uniqueIndicators.map(
    (track, index) => track.map(([key, indicator]) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_IndicatorsRenderer.IndicatorsRenderer,
      {
        ...indicator,
        trackIndex: index,
        scale: annotationsScale.axisScale,
        isInverted
      },
      key
    ))
  ) });
};
XYChartAnnotationsIndicators["displayName"] = "XyChartAnnotationsIndicators";
