// @dynatrace/strato-components/esm/layouts/container/Container.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import {
  createContext,
  forwardRef,
  useContext,
  useMemo
} from "react";
import { containerCSS } from "./Container.sty.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { containerColorsCSS } from "../../styles/container.sty.js";
import {
  getLayoutSizeStyles
} from "../../styles/getLayoutSizeStyles.js";
import {
  getSpacingSprinkles
} from "../../styles/getSpacingSprinkles.js";
const COMPONENT_NAME = "Container";
const ContainerColoringContext = createContext({
  variant: "default",
  color: "neutral",
  default: true
});
const useContainerColoring = () => {
  return useContext(ContainerColoringContext);
};
const Container = /* @__PURE__ */ forwardRef(
  ({
    as,
    children,
    variant = "default",
    color = "neutral",
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "data-dt-component": dataDtComponent = COMPONENT_NAME,
    ...remainingProps
  }, ref) => {
    const [spacingClasses, sprinkleRestProps] = getSpacingSprinkles(
      remainingProps,
      { padding: 16 }
    );
    const [sizeStyles, sizeRestProps] = getLayoutSizeStyles(sprinkleRestProps);
    const [behavioralTrackingProps, finalRestProps] = useBehavioralTrackingProps(sizeRestProps);
    const Component = as || "div";
    const context = useMemo(
      () => ({ color, variant, default: false }),
      [color, variant]
    );
    return /* @__PURE__ */ jsx(ContainerColoringContext.Provider, { value: context, children: /* @__PURE__ */ jsx(
      Component,
      {
        ref,
        className: clsx(
          containerColorsCSS({ color, variant }),
          containerCSS,
          spacingClasses,
          consumerClassName
        ),
        style: {
          ...sizeStyles,
          ...consumerStyle
        },
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        ...finalRestProps,
        ...behavioralTrackingProps,
        "data-dt-component": dataDtComponent,
        children
      }
    ) });
  }
);
Container.displayName = "Container";
export {
  Container,
  useContainerColoring
};
