// @dynatrace/strato-components/charts/gauge/components/GaugeRootSVG.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var GaugeRootSVG_exports = {};
__export(GaugeRootSVG_exports, {
  GaugeRootSVG: () => GaugeRootSVG
});
module.exports = __toCommonJS(GaugeRootSVG_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_constants = require("../constants.js");
var import_useGaugeChartSizeContext = require("../hooks/useGaugeChartSizeContext.js");
const GaugeRootSVG = ({ children }) => {
  const dimensions = (0, import_useGaugeChartSizeContext.useGaugeChartSizeContext)();
  const width = dimensions?.width ?? 0;
  const height = dimensions?.height ?? 0;
  const verticalOffset = height > width ? height * import_constants.GAUGE_VERTICAL_SMALL_OFFSET : height * import_constants.GAUGE_VERTICAL_DEFAULT_OFFSET;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "svg",
    {
      width: "100%",
      height: "100%",
      viewBox: `0 0 ${width} ${height}`,
      style: { display: "block" },
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("g", { transform: `translate(${width / 2}, ${height / 2 + verticalOffset})`, children })
    }
  );
};
