// @dynatrace/strato-components/charts/gauge/components/GaugeThresholdIndicators.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var GaugeThresholdIndicators_exports = {};
__export(GaugeThresholdIndicators_exports, {
  GaugeThresholdIndicators: () => GaugeThresholdIndicators
});
module.exports = __toCommonJS(GaugeThresholdIndicators_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_constants = require("../constants.js");
var import_useGaugeChartThresholdContext = require("../hooks/useGaugeChartThresholdContext.js");
var import_useThresholdConfigLines = require("../hooks/useThresholdConfigLines.js");
const GaugeThresholdIndicators = () => {
  const thresholds = (0, import_useGaugeChartThresholdContext.useGaugeChartThresholdContext)();
  const lines = (0, import_useThresholdConfigLines.useThresholdConfigLines)();
  return lines.map(
    (line, i) => line && // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    thresholds[i].showIndicator && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "line",
      {
        x1: line.start.x,
        y1: line.start.y,
        x2: line.end.x,
        y2: line.end.y,
        strokeWidth: import_constants.THRESHOLD_LINE_WIDTH,
        stroke: import_constants.THRESHOLD_LINE_DEFAULT_COLOR
      },
      `threshold-line-${i}`
    )
  );
};
