// @dynatrace/strato-components/charts/gauge/components/GaugeLabels.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var GaugeLabels_exports = {};
__export(GaugeLabels_exports, {
  GaugeLabels: () => GaugeLabels
});
module.exports = __toCommonJS(GaugeLabels_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
var import_typographyStyles_css = require("../../core/typographyStyles.sty.js");
var import_useDefaultLabels = require("../hooks/useDefaultLabels.js");
var import_useGaugeChartLabelsContext = require("../hooks/useGaugeChartLabelsContext.js");
var import_useGaugeChartMinMaxLabelsPosition = require("../hooks/useGaugeChartMinMaxLabelsPosition.js");
var import_GaugeLabel = require("./GaugeLabel/GaugeLabel.js");
var import_GaugeLabelsWrapper = require("./GaugeLabelsWrapper/GaugeLabelsWrapper.js");
const GaugeLabels = ({ value, prefixIcon }) => {
  const { label } = (0, import_useGaugeChartLabelsContext.useGaugeChartLabelsContext)();
  const { defaultLabel, defaultMaxLabel, defaultMinLabel, defaultValueLabel } = (0, import_useDefaultLabels.useDefaultLabels)(value);
  const { maxLabelPosition, minLabelPosition, minMaxPadding } = (0, import_useGaugeChartMinMaxLabelsPosition.useGaugeChartMinMaxLabelsPosition)();
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "g",
      {
        fill: import_colors.default.Text.Neutral.Default,
        id: "internal-labels",
        className: import_typographyStyles_css.replaceSlashedZero,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_GaugeLabelsWrapper.GaugeLabelsWrapper, { children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_GaugeLabel.GaugeLabel,
            {
              id: "value-label",
              value: defaultValueLabel,
              prefixIcon,
              responsive: true
            }
          ),
          !(0, import_lodash_es.isUndefined)(label) && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_GaugeLabel.GaugeLabel,
            {
              id: "label",
              value: (0, import_lodash_es.isNull)(label) ? defaultLabel : label,
              responsive: true
            }
          )
        ] })
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("g", { fill: import_colors.default.Text.Neutral.Default, id: "min-max-labels", children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_GaugeLabel.GaugeLabel,
        {
          value: defaultMinLabel,
          x: minLabelPosition.x,
          y: minLabelPosition.y + minMaxPadding,
          autoHide: true
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_GaugeLabel.GaugeLabel,
        {
          value: defaultMaxLabel,
          x: maxLabelPosition.x,
          y: maxLabelPosition.y + minMaxPadding,
          autoHide: true
        }
      )
    ] })
  ] });
};
