// @dynatrace/strato-components/filters/filter-field/utils/conversion/lezer-tree-conversion/parent-node/group.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var group_exports = {};
__export(group_exports, {
  convertFilterGroup: () => convertFilterGroup
});
module.exports = __toCommonJS(group_exports);
var import_statement = require("./statement.js");
var import_is_node_environment = require("../../../../../../core/utils/is-node-environment.js");
var import_decoration_marks = require("../../../../decorations/decoration-marks.js");
var import_validation_helper_functions = require("../../../validation-utils/validation-helper-functions.js");
var import_logical_operator = require("../logical-operator.js");
var import_error = require("../value/error.js");
function convertFilterGroup(node, doc, numberParser, internalValidatorMap, customTypes, parserConfig, context) {
  const { from, to } = node;
  const cursor = node.node.cursor();
  const convertedNode = {
    type: "Group",
    children: [],
    range: {
      from,
      to
    },
    explicit: false,
    explicitLogicalOperator: false,
    logicalOperator: "AND"
  };
  const errors = [];
  let hasOpeningBracket = false;
  let hasClosingBracket = false;
  let hasChildNodes = false;
  cursor.iterate((childNode) => {
    const childName = childNode.name;
    switch (childName) {
      case "Filter":
        break;
      case "FilterGroup": {
        if (node.name !== "Filter" && childNode.from === from && childNode.to === to) {
          break;
        }
        hasChildNodes = true;
        const { convertedNode: groupNode, errors: groupErrors } = convertFilterGroup(
          childNode,
          doc,
          numberParser,
          internalValidatorMap,
          customTypes,
          parserConfig,
          "FilterGroup"
        );
        convertedNode.children.push(groupNode);
        errors.push(...groupErrors);
        return false;
      }
      case "FilterStatement": {
        hasChildNodes = true;
        const { convertedNode: statementNode, errors: statementErrors } = (0, import_statement.convertFilterStatement)(
          childNode,
          doc,
          numberParser,
          internalValidatorMap,
          customTypes,
          parserConfig,
          context
        );
        convertedNode.children.push(statementNode);
        errors.push(...statementErrors);
        return false;
      }
      case "LogicalOperator": {
        hasChildNodes = true;
        const { convertedNode: operatorNode, errors: logicalOperatorErrors } = (0, import_logical_operator.convertLogicalOperator)(childNode, doc);
        if (logicalOperatorErrors.length > 0) {
          convertedNode.children.push({
            type: "Statement",
            range: operatorNode.range,
            key: operatorNode
          });
        } else {
          convertedNode.children.push(operatorNode);
        }
        errors.push(...logicalOperatorErrors);
        return false;
      }
      case "OpeningBracket":
        if (node.name !== "Filter") {
          hasOpeningBracket = true;
          convertedNode.explicit = true;
        }
        break;
      case "ClosingBracket":
        if (node.name !== "Filter") {
          hasClosingBracket = true;
          convertedNode.explicit = true;
        }
        break;
      case "\u26A0":
        if (node.name === "Filter") {
          if (childNode.node.firstChild) {
            const { convertedNode: errorNode } = (0, import_error.convertError)(
              cursor,
              childNode,
              doc
            );
            convertedNode.children.push(errorNode);
            errors.push({
              range: errorNode.range,
              message: "incompleteGroupError"
            });
          } else {
            const { convertedNode: errorNode } = (0, import_error.convertError)(
              cursor,
              childNode,
              doc
            );
            convertedNode.children.push(errorNode);
            if (parserConfig?.searchConversion) {
              errors.push({
                range: errorNode.range,
                message: {
                  type: "incompleteStatementWithEnabledSearchError",
                  details: (0, import_validation_helper_functions.getColoredSearchOperator)()
                }
              });
            } else {
              errors.push({
                range: errorNode.range,
                message: "incompleteStatementError"
              });
            }
          }
        }
        return false;
      default:
        if ((0, import_is_node_environment.isDev)()) {
          console.log(
            `${doc.toString()} | Syntax node '${childName}' cannot be converted. Please make sure to handle all nodes accordingly.`
          );
        }
    }
  });
  const groupError = {
    type: "Error",
    range: { from, to },
    value: "",
    textValue: ""
  };
  const emptyBracketGroup = hasOpeningBracket && hasClosingBracket && !hasChildNodes;
  if (emptyBracketGroup) {
    groupError.range = { from: from + 1, to: to - 1 };
    convertedNode.children.push(groupError);
    errors.push({
      range: { from, to },
      message: "emptyGroupError"
    });
  } else if (hasOpeningBracket && !hasClosingBracket) {
    groupError.range = { from: to, to };
    convertedNode.children.push(groupError);
    errors.push({
      range: { from: to - 1, to },
      message: "incompleteGroupError",
      decoration: import_decoration_marks.postSpotErrorMark
    });
  }
  return { convertedNode, errors };
}
