// @dynatrace/strato-components/filters/filter-field/utils/conversion/lezer-tree-conversion/parent-node/inclusion.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var inclusion_exports = {};
__export(inclusion_exports, {
  convertInclusionValue: () => convertInclusionValue
});
module.exports = __toCommonJS(inclusion_exports);
var import_is_node_environment = require("../../../../../../core/utils/is-node-environment.js");
var import_get_text = require("../../../get-text.js");
var import_node_mutations = require("../node-mutations.js");
var import_boolean = require("../value/boolean.js");
var import_duration = require("../value/duration.js");
var import_identifier = require("../value/identifier.js");
var import_number = require("../value/number.js");
var import_simple_value = require("../value/simple-value.js");
function convertInclusionValue(node, doc, numberParser, customTypes, parserConfig) {
  const { from, to } = node;
  const cursor = node.node.cursor();
  const convertedNode = {
    type: "List",
    range: {
      from,
      to
    },
    textValue: (0, import_get_text.getTextFromRange)(node, doc),
    value: []
  };
  const errors = [];
  const parentheses = { opening: false, closing: false };
  let highlightFullListAsError = false;
  cursor.iterate((childNode) => {
    const nodeName = childNode.name;
    const { from: childFrom, to: childTo } = childNode;
    switch (nodeName) {
      case "FilterIncludesValue":
      case "List": {
        break;
      }
      case "OpeningBracket":
        parentheses.opening = true;
        break;
      case "ClosingBracket":
        parentheses.closing = true;
        break;
      case "ListValue": {
        const {
          convertedNode: valueNode,
          errors: valueErrors,
          positionOperators
        } = (0, import_simple_value.convertSimpleValue)(childNode, doc, numberParser, customTypes, parserConfig);
        if (positionOperators) {
          (0, import_node_mutations.convertToErrorNode)(valueNode);
          for (const range of positionOperators.ranges) {
            errors.push({
              range,
              message: "invalidWildCardOperatorInList"
            });
          }
        }
        convertedNode.value.push(valueNode);
        errors.push(...valueErrors);
        return false;
      }
      case "Identifier":
      case "String":
      case "IncompleteString": {
        const {
          convertedNode: valueNode,
          positionOperators: { endsWithOperator, startsWithOperator }
        } = (0, import_identifier.convertIdentifier)(childNode, doc, nodeName);
        if (endsWithOperator) {
          (0, import_node_mutations.convertToErrorNode)(valueNode);
          errors.push({
            range: endsWithOperator,
            message: "unsupportedPositionOperator"
          });
        }
        if (startsWithOperator) {
          (0, import_node_mutations.convertToErrorNode)(valueNode);
          errors.push({
            range: startsWithOperator,
            message: "unsupportedPositionOperator"
          });
        }
        convertedNode.value.push(valueNode);
        break;
      }
      case "Number": {
        const { convertedNode: numberNode } = (0, import_number.convertNumber)(
          childNode,
          doc,
          numberParser
        );
        convertedNode.value.push(numberNode);
        break;
      }
      case "Boolean": {
        const { convertedNode: booleanNode } = (0, import_boolean.convertBoolean)(childNode, doc);
        convertedNode.value.push(booleanNode);
        break;
      }
      case "Duration": {
        const { convertedNode: durationNode } = (0, import_duration.convertDuration)(childNode, doc);
        convertedNode.value.push(durationNode);
        break;
      }
      case "PositionOperator":
      case "InvalidPositionOperator": {
        const text = (0, import_get_text.getTextFromRange)(childNode, doc);
        const errorNode = {
          type: "Error",
          range: { from: childFrom, to: childTo },
          textValue: text,
          value: text
        };
        convertedNode.value.push(errorNode);
        errors.push({
          range: { from: childFrom, to: childTo },
          message: "unsupportedPositionOperator"
        });
        break;
      }
      case "InvalidInclusionValue":
        highlightFullListAsError = true;
        (0, import_node_mutations.convertToErrorNode)(convertedNode);
        (0, import_node_mutations.updateNodeValues)(convertedNode, (0, import_get_text.getTextFromRange)(childNode, doc));
        break;
      case "\u26A0":
        highlightFullListAsError = true;
        break;
      default:
        if ((0, import_is_node_environment.isDev)()) {
          console.log(
            `${doc.toString()} | Syntax node '${childNode.name}' is not yet handled in the list conversion. Please make sure to handle all nodes accordingly.`
          );
        }
    }
  });
  if (convertedNode.value.length === 0) {
    convertedNode.textValue = (0, import_get_text.getTextFromRange)(convertedNode.range, doc);
    (0, import_node_mutations.convertToErrorNode)(convertedNode);
    errors.push({
      range: { from, to },
      message: "emptyInclusionValue"
    });
  } else if (highlightFullListAsError) {
    convertedNode.textValue = (0, import_get_text.getTextFromRange)(convertedNode.range, doc);
    (0, import_node_mutations.convertToErrorNode)(convertedNode);
    errors.push({
      range: { from, to },
      message: "invalidInclusionValue"
    });
  }
  return {
    convertedNode,
    errors
  };
}
