// @dynatrace/strato-components/filters/filter-field/utils/conversion/lezer-tree-conversion/parent-node/statement.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var statement_exports = {};
__export(statement_exports, {
  convertFilterStatement: () => convertFilterStatement
});
module.exports = __toCommonJS(statement_exports);
var import_inclusion = require("./inclusion.js");
var import_is_node_environment = require("../../../../../../core/utils/is-node-environment.js");
var import_decoration_marks = require("../../../../decorations/decoration-marks.js");
var import_error_message_helpers = require("../../../error-utils/error-message-helpers.js");
var import_get_text = require("../../../get-text.js");
var import_suggestion_type = require("../../../suggestions-utils/suggestion-type.js");
var import_validation_helper_functions = require("../../../validation-utils/validation-helper-functions.js");
var import_exists = require("../comparator/exists.js");
var import_operator = require("../comparator/operator.js");
var import_search = require("../comparator/search.js");
var import_key = require("../key/key.js");
var import_node_mutations = require("../node-mutations.js");
var import_node_type_mapping = require("../node-type-mapping.js");
var import_free_text = require("../value/free-text.js");
var import_simple_value = require("../value/simple-value.js");
function convertPositionOperators(convertedStatement, internalValidatorMap, positionOperators, errors, parserConfig) {
  if (!convertedStatement.comparisonOperator || !positionOperators) {
    return;
  }
  const operator = convertedStatement.comparisonOperator;
  const value = operator?.value;
  if (value === "=" || value === "!=" || operator?.type === "SearchOperator") {
    const newValue = `${value === "!=" ? "not-" : ""}${positionOperators.type}`;
    if (convertedStatement.comparisonOperator) {
      convertedStatement.comparisonOperator.value = newValue;
    }
    if (!(0, import_validation_helper_functions.isValidOperator)(
      newValue,
      convertedStatement.key?.textValue,
      internalValidatorMap,
      parserConfig
    )) {
      (0, import_node_mutations.convertToErrorNode)(convertedStatement.value);
      for (const positionOperator of positionOperators.ranges) {
        errors.push({
          range: positionOperator,
          message: "invalidWildCardOperator"
        });
      }
    }
  } else {
    (0, import_node_mutations.convertToErrorNode)(convertedStatement.value);
    for (const positionOperator of positionOperators.ranges) {
      errors.push({
        range: positionOperator,
        message: "invalidPositionOperator"
      });
    }
  }
}
function validateKeyAndValue(internalValidatorMap, convertedStatement, doc, errors, customTypes, parserConfig) {
  if (!internalValidatorMap || internalValidatorMap.normalizedKeys.byKey.size === 0 || !convertedStatement.key) {
    return;
  }
  if (convertedStatement.key.type !== "Error") {
    validateKey(convertedStatement.key, internalValidatorMap, errors);
  }
  if (!convertedStatement.value) {
    return;
  }
  if (Array.isArray(convertedStatement.value.value)) {
    for (const value of convertedStatement.value.value) {
      validateSingleValue(
        value,
        doc,
        { ...convertedStatement, value },
        internalValidatorMap,
        errors,
        customTypes,
        parserConfig
      );
    }
  } else {
    const isValid = validateSingleValue(
      convertedStatement.value,
      doc,
      convertedStatement,
      internalValidatorMap,
      errors,
      customTypes,
      parserConfig
    );
    if (!isValid) {
      if (convertedStatement.comparisonOperator) {
        convertedStatement.comparisonOperator.value = convertedStatement.comparisonOperator.textValue;
      }
    }
  }
}
function validateSingleValue(value, doc, convertedStatement, internalValidatorMap, errors = [], customTypes, parserConfig) {
  if (!value || !convertedStatement.key) {
    return true;
  }
  const { isValid, message, validTypes } = (0, import_validation_helper_functions.isValidValue)(
    convertedStatement.key.value,
    value.type === "String" ? value.value : value.textValue,
    value.type,
    internalValidatorMap,
    convertedStatement,
    doc.toString(),
    customTypes,
    parserConfig
  );
  if (!isValid) {
    (0, import_node_mutations.convertToErrorNode)(value);
    if (!validTypes || validTypes.length === 1) {
      const singleType = validTypes ? validTypes[0] : void 0;
      if (singleType && (0, import_error_message_helpers.isCustomType)(singleType)) {
        errors.push({
          range: value.range,
          message: message ?? {
            type: "invalidValueErrorMultipleTypes",
            details: validTypes
          }
        });
      } else {
        errors.push({
          range: value.range,
          message: message ?? (0, import_error_message_helpers.getErrorMessageForType)(singleType)
        });
      }
    } else {
      errors.push({
        range: value.range,
        message: message ?? {
          type: "invalidValueErrorMultipleTypes",
          details: validTypes
        }
      });
    }
  }
  return isValid;
}
function validateKey(key, internalValidatorMap, errors = []) {
  if (!internalValidatorMap || !key) {
    return;
  }
  const isExhaustive = internalValidatorMap.exhaustive ?? true;
  let { value } = key;
  let isValidJsonPath = true;
  if (key.type === "JSONPath") {
    value = key.root;
    isValidJsonPath = hasJSONPathTypePredicate(internalValidatorMap, value);
  }
  const keyIsInValidatorMap = internalValidatorMap.normalizedKeys.byKey.has(value);
  if ((!keyIsInValidatorMap || !isValidJsonPath) && isExhaustive) {
    (0, import_node_mutations.convertToErrorNode)(key);
    errors.push({
      range: key.range,
      message: "invalidKey"
    });
  }
}
function convertFilterStatement(node, doc, numberParser, internalValidatorMap, customTypes, parserConfig, context) {
  const nodeName = node.name;
  const { from, to } = node;
  const cursor = node.node.cursor();
  const convertedStatement = {
    type: import_node_type_mapping.nodeTypeOverrides[nodeName] ?? nodeName,
    range: {
      from,
      to
    }
  };
  const errors = [];
  let positionOperators;
  let prevNode;
  function updatePreviousAndContinue(nodeName2) {
    prevNode = nodeName2;
    return false;
  }
  cursor.iterate((childNode) => {
    const nodeName2 = childNode.name;
    const { from: childFrom, to: childTo } = childNode;
    switch (nodeName2) {
      case "FilterStatement": {
        break;
      }
      case "FilterKey": {
        const { convertedNode: filterKey, errors: keyErrors } = (0, import_key.convertFilterKey)(childNode, doc, parserConfig);
        if (filterKey !== void 0) {
          convertedStatement.key = filterKey;
        }
        errors.push(...keyErrors);
        return updatePreviousAndContinue(nodeName2);
      }
      case "InvalidComparisonOperator":
      case "ComparisonOperator":
      case "InclusionOperator": {
        const { convertedNode: operator2 } = (0, import_operator.convertComparisonOperator)(
          childNode,
          doc
        );
        convertedStatement.comparisonOperator = operator2;
        return updatePreviousAndContinue(nodeName2);
      }
      case "SearchOperator": {
        const { convertedNode: operator2, errors: operatorErrors } = (0, import_search.convertSearchOperator)(childNode, doc, parserConfig, context);
        convertedStatement.comparisonOperator = operator2;
        errors.push(...operatorErrors);
        break;
      }
      case "ExistsOperator": {
        const { convertedNode: operator2 } = (0, import_exists.convertExistsOperator)(
          childNode,
          doc
        );
        convertedStatement.comparisonOperator = operator2;
        break;
      }
      case "FilterSimpleValue": {
        const {
          convertedNode: value,
          errors: valueErrors,
          positionOperators: valuePositionOperators
        } = (0, import_simple_value.convertSimpleValue)(
          childNode,
          doc,
          numberParser,
          customTypes,
          parserConfig
        );
        convertedStatement.value = value;
        errors.push(...valueErrors);
        positionOperators = valuePositionOperators;
        return updatePreviousAndContinue(nodeName2);
      }
      case "FilterIncludesValue": {
        const { convertedNode: value, errors: valueErrors } = (0, import_inclusion.convertInclusionValue)(
          childNode,
          doc,
          numberParser,
          customTypes,
          parserConfig
        );
        convertedStatement.value = value;
        errors.push(...valueErrors);
        return updatePreviousAndContinue(nodeName2);
      }
      case "FreeText": {
        const { convertedNode: freeTextNode, errors: freeTextErrors } = (0, import_free_text.convertFreeText)(childNode, doc, parserConfig);
        convertedStatement.key = freeTextNode;
        errors.push(...freeTextErrors);
        return updatePreviousAndContinue(nodeName2);
      }
      // An error node may occur as direct child of a 'FilterStatement'.
      // Depending on the context, it needs to be added to the corresponding node (key, comparator, or value).
      case "\u26A0":
        switch (prevNode) {
          case "FilterKey":
          case "FreeText":
            if (parserConfig?.searchConversion) {
              errors.push({
                range: { from: childFrom - 1, to: childTo },
                message: {
                  type: "incompleteStatementWithEnabledSearchError",
                  details: (0, import_validation_helper_functions.getColoredSearchOperator)()
                },
                decoration: import_decoration_marks.postSpotErrorMark
              });
            } else {
              errors.push({
                range: { from: childFrom - 1, to: childTo },
                message: "incompleteStatementError",
                decoration: import_decoration_marks.postSpotErrorMark
              });
            }
            break;
          case "ComparisonOperator":
          case "InvalidComparisonOperator":
          case "SearchOperator":
            if (convertedStatement.comparisonOperator !== void 0) {
              if (childFrom !== childTo) {
                const childAfter = childNode.node.nextSibling;
                if (childAfter) {
                  if (convertedStatement.comparisonOperator) {
                    convertedStatement.comparisonOperator.range.to = childTo;
                    const textValue = (0, import_get_text.getTextFromRange)(
                      convertedStatement.comparisonOperator.range,
                      doc
                    );
                    convertedStatement.comparisonOperator.textValue = convertedStatement.comparisonOperator.value = textValue;
                  }
                  return updatePreviousAndContinue(prevNode);
                } else {
                  const range = {
                    from: childFrom,
                    to: childTo
                  };
                  const textValue = (0, import_get_text.getTextFromRange)(range, doc);
                  convertedStatement.value = {
                    type: "Error",
                    value: textValue,
                    textValue,
                    range
                  };
                  errors.push({
                    range,
                    message: "invalidValueErrorGeneric"
                  });
                }
              } else {
                errors.push({
                  range: {
                    from: childFrom - (childFrom === childTo ? 1 : 0),
                    to: childTo
                  },
                  message: "missingValueError",
                  decoration: import_decoration_marks.postSpotErrorMark
                });
              }
            }
            break;
          case "InclusionOperator":
            errors.push({
              range: { from: childFrom - 1, to: childTo },
              message: "missingValueError"
            });
            break;
          default:
            if ((0, import_is_node_environment.isDev)()) {
              console.log(
                `${doc.toString()} | Syntax node '${nodeName2}' is not yet handled after a '${prevNode}' node. Please make sure to handle all nodes accordingly.`
              );
            }
        }
        return updatePreviousAndContinue(nodeName2);
      default:
        console.log(
          `${doc.toString()} | Syntax node '${nodeName2}' is not yet handled in the statement conversion. Please make sure to handle all nodes accordingly.`
        );
    }
    prevNode = nodeName2;
  });
  if (convertedStatement.comparisonOperator && positionOperators) {
    convertPositionOperators(
      convertedStatement,
      internalValidatorMap,
      positionOperators,
      errors,
      parserConfig
    );
  }
  validateKeyAndValue(
    internalValidatorMap,
    convertedStatement,
    doc,
    errors,
    customTypes,
    parserConfig
  );
  const operator = convertedStatement.comparisonOperator;
  if (operator && !positionOperators && (!(0, import_validation_helper_functions.isValidOperator)(
    operator.textValue,
    convertedStatement.key?.value,
    internalValidatorMap,
    parserConfig
  ) || operator.type === "Error" && operator.value !== "*~")) {
    (0, import_node_mutations.convertToErrorNode)(operator);
    if (convertedStatement.comparisonOperator) {
      (0, import_node_mutations.convertToErrorNode)(convertedStatement.comparisonOperator);
    }
    errors.push({
      range: operator.range,
      message: {
        type: "invalidComparatorError",
        details: (0, import_validation_helper_functions.getValidOperatorsColored)(
          convertedStatement.key?.textValue,
          internalValidatorMap,
          parserConfig
        )
      }
    });
  }
  return { convertedNode: convertedStatement, errors };
}
function hasJSONPathTypePredicate(internalValidatorMap, key) {
  const config = internalValidatorMap.normalizedKeys.byKey.get(key);
  if (!config) {
    return false;
  }
  if (config.valueType) {
    const types = Array.isArray(config.valueType) ? config.valueType : [config.valueType];
    return types.includes("JSONPath");
  }
  const validValues = config.valuePredicate;
  if (!validValues || typeof validValues === "function") {
    return false;
  }
  const { types: validTypes } = (0, import_suggestion_type.getValidValueTypes)(validValues);
  return validTypes.includes("JSONPath");
}
