// @dynatrace/strato-components/esm/editors/dql-editor/actions-menu/ActionsMenu.js
import { jsx, jsxs } from "react/jsx-runtime";
import { useCallback } from "react";
import { useIntl } from "react-intl";
import {
  CopyIcon,
  DotMenuIcon,
  ExternalLinkIcon
} from "@dynatrace/strato-icons";
import { dqlEditorActionsMenuMessages } from "./actions-menu-messages.js";
import * as styles from "./ActionsMenu.sty.js";
import { Button } from "../../../buttons/button/Button.js";
import { DQL_DOCUMENTATION_URL } from "../../../core/external-links.js";
import { Menu } from "../../../navigation/menu/Menu.js";
const ActionsMenu = () => null;
ActionsMenu.displayName = "DQLEditor.ActionsMenu";
const ActionsMenuInternal = ({
  getValue,
  size
}) => {
  const intl = useIntl();
  const copyToClipboard = useCallback(async () => {
    try {
      await navigator.clipboard.writeText(getValue());
    } catch {
    }
  }, [getValue]);
  return /* @__PURE__ */ jsx("div", { className: styles.actionsMenuWrapper({ size }), children: /* @__PURE__ */ jsxs(Menu, { children: [
    /* @__PURE__ */ jsx(Menu.Trigger, { children: /* @__PURE__ */ jsx(
      Button,
      {
        variant: "default",
        size,
        "aria-label": intl.formatMessage(
          dqlEditorActionsMenuMessages.triggerLabel
        ),
        "data-dt-name": "OpenActionsMenu",
        children: /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(DotMenuIcon, {}) })
      }
    ) }),
    /* @__PURE__ */ jsxs(Menu.Content, { children: [
      /* @__PURE__ */ jsx(Menu.Group, { children: /* @__PURE__ */ jsxs(
        Menu.Item,
        {
          onSelect: () => {
            copyToClipboard().catch(() => void 0);
          },
          "data-dt-name": "CopyQuery",
          children: [
            /* @__PURE__ */ jsx(Menu.Prefix, { children: /* @__PURE__ */ jsx(CopyIcon, {}) }),
            intl.formatMessage(dqlEditorActionsMenuMessages.copyQuery)
          ]
        }
      ) }),
      /* @__PURE__ */ jsx(Menu.Group, { children: /* @__PURE__ */ jsxs(
        Menu.Link,
        {
          href: DQL_DOCUMENTATION_URL,
          target: "_blank",
          rel: "noopener noreferrer",
          "data-dt-name": "OpenDocumentation",
          children: [
            /* @__PURE__ */ jsx(Menu.Prefix, { children: /* @__PURE__ */ jsx(ExternalLinkIcon, {}) }),
            intl.formatMessage(
              dqlEditorActionsMenuMessages.openDocumentation
            )
          ]
        }
      ) })
    ] })
  ] }) });
};
export {
  ActionsMenu,
  ActionsMenuInternal
};
