// @dynatrace/strato-components/esm/filters/filter-field/utils/error-utils/error-message-helpers.js
const builtInTypeLabelKeys = {
  Duration: "Durations",
  Number: "Numbers",
  String: "Strings",
  IPAddress: "IPAddresses",
  UID: "UIDs",
  Timestamp: "Timestamps",
  SmartscapeId: "SmartscapeIds",
  Boolean: "Boolean"
};
function getTypeLabelKey(type) {
  return builtInTypeLabelKeys[type];
}
function getCustomTypeLabel(type, customTypes) {
  if (!customTypes) {
    return void 0;
  }
  const customType = customTypes[type];
  if (customType && typeof customType === "object" && customType.label) {
    return customType.label;
  }
  return void 0;
}
function isCustomType(type) {
  return !builtInTypeLabelKeys[type] && type !== "Any" && type !== "JSONPath";
}
function getErrorMessageForType(type) {
  switch (type) {
    case "Boolean":
      return "invalidValueErrorBoolean";
    case "Duration":
      return "invalidValueErrorDurations";
    case "Number":
      return "invalidValueErrorNumbers";
    case "IPAddress":
      return "invalidValueErrorIPAddresses";
    case "UID":
      return "invalidValueErrorUIDs";
    case "SmartscapeId":
      return "invalidValueErrorSmartscapeIds";
    case "Timestamp":
      return "invalidValueErrorTimestamps";
    case "String":
      return "invalidValueErrorStrings";
    case "Any":
    default:
      return "invalidValueErrorGeneric";
  }
}
export {
  getCustomTypeLabel,
  getErrorMessageForType,
  getTypeLabelKey,
  isCustomType
};
