// @dynatrace/strato-components/esm/tables/DataTable/features/GanttChart/DataTableGanttChartHeader.js
import { jsx, jsxs } from "react/jsx-runtime";
import { isEmpty } from "lodash-es";
import { useCallback, useRef } from "react";
import * as styles from "./DataTableGanttChartHeader.sty.js";
import { useMinMaxFromConfig } from "./hooks/useMinMaxFromConfig.js";
import { useScaleAndFormatter } from "./hooks/useScaleAndFormatter.js";
import { toIntegrationTracks } from "./to-integration-tracks.js";
import { AnnotationsNumericalDisplay } from "../../../../charts/core/components/annotations/components/displays/AnnotationsNumericalDisplay.js";
import { AnnotationsTimeDisplay } from "../../../../charts/core/components/annotations/components/displays/AnnotationsTimeDisplay.js";
import { AnnotationsScrollWrapper } from "../../../../charts/core/components/annotations/components/overlays/AnnotationsScrollWrapper.js";
import {
  TRACK_HEIGHT,
  TRACK_SEPARATOR_STROKE_WIDTH
} from "../../../../charts/core/components/annotations/constants.js";
import { TimeTableHeaderAxis } from "../../../../charts/core/components/axes/components/TimeTableHeaderAxis.js";
import { ValueTableHeaderAxis } from "../../../../charts/core/components/axes/components/ValueTableHeaderAxis.js";
import { ChartStateHandler } from "../../../../charts/core/components/chart-states/ChartStateHandler.js";
import { coerceSizeValue } from "../../../../charts/core/utils/coerceSizeValue.js";
import { useContainerSize } from "../../../../core/hooks/useContainerSize.js";
import { Flex } from "../../../../layouts/flex/Flex.js";
import { DataTableDefaultCell } from "../../components/DataTableDefaultCell.js";
import {
  DATA_TABLE_CELL_HORIZONTAL_PADDING_MAP as CELL_HORIZONTAL_PADDING_MAP,
  DATA_TABLE_DEFAULT_ICON_PADDING as ICON_PADDING,
  DATA_TABLE_SUB_ROW_EXPANDER_ICON_SIZE as SUB_ROW_EXPANDER_ICON_SIZE,
  DATA_TABLE_SUB_ROW_INDENT_GRID as SUB_ROW_INDENT_GRID
} from "../../constants.js";
import { useDataTable } from "../../contexts/DataTableContext.js";
const DataTableGanttChartHeader = ({
  columnDef
}) => {
  const sizeRef = useRef(null);
  const { width: axisWidth, height: axisHeight } = useContainerSize(sizeRef);
  const config = columnDef.config || {};
  const { xAxisType = "numerical", annotationsHeader } = config;
  const { min, max } = useMinMaxFromConfig(columnDef.id);
  const table = useDataTable();
  const isSubrowIndicatorColumn = table.getSubRowIndicatorColumnId() === columnDef.id;
  const rowDensity = table.getVisualVariantOptions().rowDensity;
  const leftHeaderPaddingValue = isSubrowIndicatorColumn ? SUB_ROW_INDENT_GRID * table.getSubRowOpenDepth() : 0;
  const leftHeaderPadding = isSubrowIndicatorColumn ? `${leftHeaderPaddingValue}px` : "inherit";
  const subrowExpanderWidth = isSubrowIndicatorColumn ? SUB_ROW_EXPANDER_ICON_SIZE + 2 * ICON_PADDING + CELL_HORIZONTAL_PADDING_MAP[rowDensity] : 0;
  const effectiveAxisWidth = Math.max(
    0,
    axisWidth - leftHeaderPaddingValue - subrowExpanderWidth
  );
  const { scale, formatter } = useScaleAndFormatter(
    effectiveAxisWidth,
    config,
    min,
    max
  );
  const {
    tooltip: annotationsTooltip,
    tooltipActions: annotationsTooltipActions,
    emptyState,
    errorState,
    loading,
    textOverflow = "truncate",
    truncateMode = "middle",
    tracks: annotationTracksConfig,
    height: annotationsHeightConfig
  } = annotationsHeader ?? {};
  const tooltipConfig = annotationsTooltip ? { template: annotationsTooltip } : {};
  const columnId = columnDef.id;
  const handleMarkerGroupHover = useCallback(
    (event) => {
      table.setGanttAnnotationsSelection(columnId, event);
    },
    [table, columnId]
  );
  const annotationTracks = annotationTracksConfig ? toIntegrationTracks(annotationTracksConfig) : [];
  const shownTracks = annotationTracks.filter(({ hidden }) => !hidden);
  const tracksHeight = shownTracks.length * TRACK_HEIGHT + TRACK_SEPARATOR_STROKE_WIDTH;
  const configuredHeight = coerceSizeValue(annotationsHeightConfig);
  const scrollWrapperHeight = configuredHeight ? `min(${tracksHeight}px, ${configuredHeight})` : tracksHeight;
  let ganttAxis = /* @__PURE__ */ jsx(
    ValueTableHeaderAxis,
    {
      scale,
      formatter,
      containerHeight: axisHeight,
      containerWidth: effectiveAxisWidth,
      min,
      max
    }
  );
  let annotationsDisplay = /* @__PURE__ */ jsx(
    AnnotationsNumericalDisplay,
    {
      scale,
      tracks: annotationTracks,
      showLabels: false,
      labelsOffset: 0,
      width: effectiveAxisWidth,
      markerGroupHover: handleMarkerGroupHover,
      tooltipConfig,
      annotationsActions: annotationsTooltipActions,
      textOverflow,
      truncateMode,
      formatter
    }
  );
  if (xAxisType === "time") {
    ganttAxis = /* @__PURE__ */ jsx(
      TimeTableHeaderAxis,
      {
        scale,
        containerHeight: axisHeight,
        containerWidth: effectiveAxisWidth,
        min,
        max
      }
    );
    annotationsDisplay = /* @__PURE__ */ jsx(
      AnnotationsTimeDisplay,
      {
        scale,
        tracks: annotationTracks,
        showLabels: false,
        labelsOffset: 0,
        width: effectiveAxisWidth,
        markerGroupHover: handleMarkerGroupHover,
        tooltipConfig,
        annotationsActions: annotationsTooltipActions,
        textOverflow,
        truncateMode
      }
    );
  }
  return /* @__PURE__ */ jsx(
    DataTableDefaultCell,
    {
      ref: sizeRef,
      className: styles.ganttChartHeader,
      "data-sizeref": true,
      children: /* @__PURE__ */ jsxs(
        Flex,
        {
          flexDirection: "column",
          minHeight: annotationsHeightConfig,
          justifyContent: "space-between",
          style: { paddingLeft: leftHeaderPadding },
          children: [
            annotationsHeader && /* @__PURE__ */ jsx(
              ChartStateHandler,
              {
                empty: isEmpty(annotationTracks),
                height: "100%",
                emptyState,
                errorState,
                isLoading: loading ?? false,
                style: { margin: "auto" },
                children: /* @__PURE__ */ jsx(
                  AnnotationsScrollWrapper,
                  {
                    height: scrollWrapperHeight,
                    width: effectiveAxisWidth,
                    children: annotationsDisplay
                  }
                )
              }
            ),
            ganttAxis
          ]
        }
      )
    }
  );
};
export {
  DataTableGanttChartHeader
};
