// @dynatrace/strato-components/esm/tables/DataTable/features/GanttChart/DataTableGanttAnnotationsIndicators.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import * as styles from "./DataTableGanttAnnotationsIndicators.sty.js";
import { DATA_TABLE_ROW_BORDER } from "../../constants.js";
import { useScaleAndFormatter } from "./hooks/useScaleAndFormatter.js";
import { toIntegrationTracks } from "./to-integration-tracks.js";
import { AreaIndicator } from "../../../../charts/core/components/annotations/components/indicators/AreaIndicator.js";
import { LineIndicator } from "../../../../charts/core/components/annotations/components/indicators/LineIndicator.js";
import { getShownIndicators } from "../../../../charts/core/components/annotations/components/indicators/utils/get-shown-indicators.js";
import { getUniqueIndicatorsByPriority } from "../../../../charts/core/components/annotations/components/indicators/utils/get-unique-indicators-by-priority.js";
import { getMarkerBrightnessBasedOnState } from "../../../../charts/core/components/annotations/utils/get-marker-brightness-based-on-state.js";
import { isRangeMarker } from "../../../../charts/core/components/annotations/utils/is-range-marker.js";
import { serializeMarkerRange } from "../../../../charts/core/components/annotations/utils/serialize-marker-range.js";
import { BrightnessFilter } from "../../../../charts/core/components/BrightnessFilter.js";
import { useCurrentTheme } from "../../../../core/hooks/useCurrentTheme.js";
import { useDataTable } from "../../contexts/DataTableContext.js";
function getBrightnessFilterId(indicator, trackIndex) {
  const { start, end } = indicator;
  return `gantt_brightness_${serializeMarkerRange(
    {
      start: start.valueOf(),
      end: end ? end.valueOf() : void 0
    },
    trackIndex
  )}`;
}
function useGanttIndicators(config, columnId, range) {
  const table = useDataTable();
  const { tracks: configuredTracks } = config.annotationsHeader ?? {};
  if (!configuredTracks?.length) {
    return null;
  }
  const selection = table.getState().ganttAnnotationsSelection?.[columnId] ?? {};
  const tracks = toIntegrationTracks(configuredTracks);
  const shownIndicators = getShownIndicators(tracks, selection);
  const uniqueIndicators = getUniqueIndicatorsByPriority(
    shownIndicators,
    range
  );
  const hasIndicators = uniqueIndicators.some((track) => track.length > 0);
  return hasIndicators ? uniqueIndicators : null;
}
function GanttIndicatorShape({
  indicator,
  trackIndex,
  scale,
  height,
  theme
}) {
  const { start, end, color, state } = indicator;
  const brightnessFilterId = getBrightnessFilterId(indicator, trackIndex);
  const slope = getMarkerBrightnessBasedOnState(theme, state);
  const filter = `url(#${brightnessFilterId})`;
  const range = { start, end };
  const x = scale(start) ?? 0;
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(BrightnessFilter, { id: brightnessFilterId, slope }),
    isRangeMarker(range) ? /* @__PURE__ */ jsx(
      AreaIndicator,
      {
        x,
        width: (scale(range.end) ?? 0) - x,
        height,
        color,
        filter
      }
    ) : /* @__PURE__ */ jsx(LineIndicator, { x, height, color, filter })
  ] });
}
const DataTableGanttAnnotationIndicators = ({
  width,
  height,
  min,
  max,
  config,
  columnId,
  isLastPageRow
}) => {
  const theme = useCurrentTheme();
  const { scale } = useScaleAndFormatter(width, config, min, max);
  const indicators = useGanttIndicators(config, columnId, [min, max]);
  if (!indicators) {
    return null;
  }
  const fullRowHeight = height + (isLastPageRow ? 0 : DATA_TABLE_ROW_BORDER);
  return /* @__PURE__ */ jsx(
    "svg",
    {
      height: fullRowHeight,
      width,
      className: styles.annotationIndicators,
      children: indicators.map(
        (track, trackIndex) => track.map(([key, indicator]) => /* @__PURE__ */ jsx(
          GanttIndicatorShape,
          {
            indicator,
            trackIndex,
            scale,
            height: fullRowHeight,
            theme
          },
          key
        ))
      )
    }
  );
};
export {
  DataTableGanttAnnotationIndicators
};
