// @dynatrace/strato-components/esm/tables/DataTable/features/GanttChart/DataTableGanttChartGrid.js
import { jsx } from "react/jsx-runtime";
import { getTicks } from "@visx/scale";
import { useMemo } from "react";
import borders from "@dynatrace/strato-design-tokens/borders";
import colors from "@dynatrace/strato-design-tokens/colors";
import * as styles from "./DataTableGanttChartGrid.sty.js";
import { useScaleAndFormatter } from "./hooks/useScaleAndFormatter.js";
import { getTickValueAsNumber } from "../../../../charts/core/utils/get-tick-value-as-number.js";
import {
  getGanttHorizontalTimeAxisNumTicks,
  getHorizontalNumTicks
} from "../../../../charts/core/utils/ticks.js";
const DataTableGanttChartGrid = ({
  width,
  height,
  min,
  max,
  config,
  isSubrowIndicatorColumn
}) => {
  const { xAxisType = "numerical" } = config;
  const { scale } = useScaleAndFormatter(width, config, min, max);
  const numTicks = xAxisType === "time" ? getGanttHorizontalTimeAxisNumTicks(width) : getHorizontalNumTicks(width);
  const lines = useMemo(() => {
    const ticks = getTicks(scale, numTicks);
    if (!isSubrowIndicatorColumn && getTickValueAsNumber(ticks.at(0)?.valueOf()) === min) {
      ticks.shift();
    }
    if (getTickValueAsNumber(ticks.at(-1)?.valueOf()) === max) {
      ticks.pop();
    }
    return ticks.map((rawTick) => {
      const tick = Math.max(scale(rawTick), 0.5);
      return /* @__PURE__ */ jsx(
        "line",
        {
          x1: tick,
          x2: tick,
          y1: 0,
          y2: height,
          strokeWidth: borders.Width.Default,
          stroke: colors.Border.Neutral.Default
        },
        +rawTick
      );
    });
  }, [scale, numTicks, min, max, height]);
  return /* @__PURE__ */ jsx(
    "svg",
    {
      viewBox: `0 0 ${width} ${height}`,
      height,
      width,
      className: styles.ganttChartGrid,
      children: lines
    }
  );
};
export {
  DataTableGanttChartGrid
};
