// @dynatrace/strato-components/esm/tables/DataTable/features/GanttChart/GanttChart.js
import { useMemo } from "react";
import {
  resolveMinMax,
  resolveTimeMinMax
} from "../../../../charts/core/utils/get-min-max-for-scale.js";
import {
  isAccessorFnColumnDef,
  isAccessorKeyColumnDef,
  isDisplayColumnDef
} from "../../hooks/useTable/datatable-typeguards.js";
import { isDatatableGanttColumnConfig } from "../ColumnTypes/ColumnTypes.js";
function isDataTableGanttData(obj) {
  return typeof obj === "object" && obj !== null && "start" in obj;
}
function getDataFromRow(row, columnDef) {
  if (isAccessorKeyColumnDef(columnDef)) {
    return row[columnDef.accessorKey];
  }
  if (isAccessorFnColumnDef(columnDef)) {
    return columnDef.accessorFn(row, 0);
  }
}
const adjustMinMaxWithAnnotationsHeader = (tracks, min, max) => {
  tracks.forEach((track) => {
    track.markers.forEach((marker) => {
      const start = marker.start;
      const end = marker.end ?? marker.start;
      const startValue = start instanceof Date ? start.valueOf() : start;
      const endValue = end instanceof Date ? end.valueOf() : end;
      min = Math.min(min, startValue);
      max = Math.max(max, endValue);
    });
  });
  return { min, max };
};
function getMinMaxForGanttColumns(data, columns, getSubRows) {
  const minMaxArray = columns.filter(
    (columnDef) => isDisplayColumnDef(columnDef) && columnDef.columnType === "gantt"
  ).map((columnDef) => {
    const config = columnDef.meta?.config;
    let configMin = "data-min";
    let configMax = "data-max";
    let xAxisType = "numerical";
    let annotationsHeader;
    if (isDatatableGanttColumnConfig(config)) {
      configMin = config.min ?? configMin;
      configMax = config.max ?? configMax;
      xAxisType = config.xAxisType ?? xAxisType;
      annotationsHeader = config.annotationsHeader;
    }
    let ganttMin = Number.POSITIVE_INFINITY;
    let ganttMax = Number.NEGATIVE_INFINITY;
    const findGanttMinMax = (row) => {
      const ganttData = getDataFromRow(row, columnDef);
      const ganttSegments = Array.isArray(ganttData) ? ganttData : [ganttData];
      ganttSegments.forEach((segment) => {
        if (isDataTableGanttData(segment)) {
          const { start, end } = segment;
          ganttMin = Math.min(ganttMin, start);
          ganttMax = Math.max(ganttMax, end ?? start);
        }
      });
      const subRows = getSubRows?.(row, 0);
      subRows?.forEach((subRow) => findGanttMinMax(subRow));
    };
    for (const row of data) {
      findGanttMinMax(row);
    }
    if (annotationsHeader) {
      ({ min: ganttMin, max: ganttMax } = adjustMinMaxWithAnnotationsHeader(
        annotationsHeader.tracks,
        ganttMin,
        ganttMax
      ));
    }
    return [
      columnDef.id,
      xAxisType === "numerical" ? resolveMinMax(
        { min: ganttMin, max: ganttMax },
        { min: configMin, max: configMax }
      ) : resolveTimeMinMax(
        { min: ganttMin, max: ganttMax },
        { min: configMin, max: configMax }
      )
    ];
  });
  return Object.fromEntries(minMaxArray);
}
const DataTableGanttChartFeature = {
  createTable(table) {
    table.setGanttAnnotationsSelection = (columnId, selection) => {
      table.setState((state) => ({
        ...state,
        ganttAnnotationsSelection: {
          ...state.ganttAnnotationsSelection,
          [columnId]: {
            ...state.ganttAnnotationsSelection?.[columnId],
            ...selection
          }
        }
      }));
    };
  }
};
function useGanttChart(_props, options) {
  const ganttMinMaxValues = useMemo(
    () => getMinMaxForGanttColumns(
      options.data,
      options.columns,
      options.getSubRows
    ),
    [options.columns, options.data, options.getSubRows]
  );
  options.state = {
    ...options.state,
    ganttMinMaxValues
  };
}
export {
  DataTableGanttChartFeature,
  getMinMaxForGanttColumns,
  useGanttChart
};
