// @dynatrace/strato-components/charts/annotations-chart/utils/annotations-chart-config.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var annotations_chart_config_exports = {};
__export(annotations_chart_config_exports, {
  buildAnnotationsChartConfig: () => buildAnnotationsChartConfig,
  deserializeConfig: () => deserializeConfig,
  isSerializedConfig: () => isSerializedConfig
});
module.exports = __toCommonJS(annotations_chart_config_exports);
var import_lodash_es = require("lodash");
var import_populate_missing_tooltip_props = require("./populate-missing-tooltip-props.js");
var import_constants = require("../../core/components/annotations/constants.js");
var import_populate_missing_axis_props = require("../components/axis/utils/populate-missing-axis-props.js");
var import_constants2 = require("../constants.js");
function deserializeConfig(serializedConfig) {
  try {
    return (0, import_lodash_es.merge)(
      {},
      import_constants2.DEFAULT_ANNOTATIONS_CHART_CONFIG,
      JSON.parse(serializedConfig)
    );
  } catch {
    return import_constants2.DEFAULT_ANNOTATIONS_CHART_CONFIG;
  }
}
function isSerializedConfig(contextConfig) {
  return typeof contextConfig === "string";
}
function buildAnnotationsChartConfig(props, contextConfig, xAxis, tooltip, hasNumericalData) {
  return {
    showLabels: props.showLabels ?? contextConfig.showLabels ?? import_constants2.DEFAULT_ANNOTATIONS_CHART_CONFIG.showLabels,
    labelWidth: props.labelWidth ?? contextConfig.labelWidth ?? import_constants.DEFAULT_LABELS_WIDTH,
    height: props.height ?? contextConfig.height,
    textOverflow: props.textOverflow ?? contextConfig?.textOverflow ?? import_constants.DEFAULT_TEXT_OVERFLOW,
    truncateMode: props.truncateMode ?? contextConfig?.truncateMode ?? import_constants.DEFAULT_TRUNCATION_MODE,
    maxLines: props.maxLines ?? contextConfig?.maxLines ?? import_constants.DEFAULT_TOOLTIP_MAX_LINES,
    xAxisConfig: (0, import_populate_missing_axis_props.populateMissingAxisProps)(
      xAxis,
      hasNumericalData,
      contextConfig.xAxisConfig
    ),
    tooltipConfig: (0, import_populate_missing_tooltip_props.populateMissingTooltipProps)(
      tooltip,
      contextConfig.tooltipConfig
    )
  };
}
