// @dynatrace/strato-components/filters/segment-selector/SegmentSelector.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SegmentSelector_exports = {};
__export(SegmentSelector_exports, {
  SegmentSelector: () => SegmentSelector
});
module.exports = __toCommonJS(SegmentSelector_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_react_is = require("react-is");
var import_FocusScope = require("../../core/components/focus-scope/FocusScope.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useId = require("../../core/hooks/useId.js");
var import_useSyncedFormControlProps = require("../../forms/forms-core/hooks/useSyncedFormControlProps.js");
var import_formfield_messages_state_reducer = require("../../forms/forms-core/state/formfield-messages-state-reducer.js");
var import_SegmentsCache = require("./contexts/SegmentsCache/SegmentsCache.js");
var import_SegmentsTriggerContext = require("./contexts/SegmentsUI/SegmentsTriggerContext.js");
var import_SegmentsUIContext = require("./contexts/SegmentsUI/SegmentsUIContext.js");
var import_segmentValidationHelper = require("./contexts/SegmentsUI/segmentValidationHelper.js");
var import_CustomTrigger = require("./CustomTrigger.js");
var import_useSegments = require("./hooks/useSegments.js");
var import_useTooltipHandler = require("./hooks/useTooltipHandler.js");
var import_useTriggerProps = require("./hooks/useTriggerProps.js");
var import_Overlay = require("./parts/Overlay.js");
var import_SegmentSelectorTrigger = require("./parts/selectionBar/SegmentSelectorTrigger.js");
var import_TooltipMessages = require("./parts/TooltipMessages.js");
var import_segmentMessages = require("./segmentMessages.js");
var styles = __toESM(require("./SegmentSelector.sty.js"));
var import_OverlayContainer = require("../../core/components/OverlayContainer.js");
var import_VisuallyHidden = require("../../core/components/visually-hidden/VisuallyHidden.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_sdk_logger = require("../../core/sdk/sdk-logger.js");
var import_StratoLoggerContext = require("../../core/sdk/StratoLoggerContext.js");
var import_is_node_environment = require("../../core/utils/is-node-environment.js");
var import_merge_props = require("../../core/utils/merge-props.js");
var import_FormFieldMessageContextProvider = require("../../forms/forms-core/components/form-field/FormFieldMessageContextProvider.js");
var import_useFormFieldMessagesContext = require("../../forms/forms-core/hooks/useFormFieldMessagesContext.js");
var import_useFormFieldValidation = require("../../forms/forms-core/hooks/useFormFieldValidation.js");
var import_useOverlayWithTrigger = require("../../overlays/hooks/useOverlayWithTrigger.js");
var import_Overlay_css = require("../../overlays/overlay/Overlay.sty.js");
var import_Tooltip = require("../../overlays/tooltip/Tooltip.js");
const COMPONENT_NAME = "SegmentSelector";
function getCustomTriggerSlot(children) {
  let customTrigger;
  import_react.Children.forEach(children, (child) => {
    if ((0, import_react_is.isElement)(child)) {
      switch (child.type) {
        case import_CustomTrigger.CustomTrigger:
          customTrigger = child;
          break;
        default: {
          if ((0, import_is_node_environment.isDev)()) {
            console.warn(
              "Only CustomTrigger is a valid child of the SegmentSelector."
            );
          }
          break;
        }
      }
    } else if (child && (0, import_is_node_environment.isDev)()) {
      console.warn("The used child is not supported by the SegmentSelector.");
    }
  });
  return customTrigger;
}
const SegmentSelectorFwd = (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    className: consumerClassName,
    style: consumerStyle,
    id: propId,
    "data-testid": dataTestId,
    "data-dtrum-allow": dataDtrumAllow,
    "data-dtrum-mask": dataDtrumMask,
    variant = "default",
    onOpenChange,
    readOnly: propReadOnly,
    disabled: propDisabled,
    required: propRequired,
    name,
    ...remainingProps
  } = props;
  const intl = (0, import_react_intl.useIntl)();
  const inputRef = (0, import_react.useRef)(null);
  const internalButtonRef = (0, import_react.useRef)(null);
  const { disabled, required, id, ariaDisabled, readOnly } = (0, import_useSyncedFormControlProps.useSyncedFormControlProps)({
    propDisabled,
    propRequired,
    propId,
    propAriaDisabled: remainingProps["aria-disabled"],
    propReadOnly
  });
  const {
    isLoading,
    text,
    ariaProps: triggerAriaProps
  } = (0, import_useTriggerProps.useTriggerProps)(variant, remainingProps);
  const [behavioralTrackingProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const { onChange, onInvalid, ariaProps } = (0, import_useFormFieldValidation.useFormFieldValidation)(
    inputRef,
    { required },
    triggerAriaProps,
    id
  );
  const { state: formFieldMessagesState, dispatch: formFieldMessagesDispatch } = (0, import_useFormFieldMessagesContext.useFormFieldMessagesContext)();
  const errorId = (0, import_useId.useId)();
  const { segments } = (0, import_useSegments.useSegments)();
  const stringifiedValue = JSON.stringify(segments);
  const segmentSelectorProps = {
    "data-dtrum-allow": dataDtrumAllow,
    "data-dtrum-mask": dataDtrumMask,
    "data-testid": dataTestId,
    style: consumerStyle,
    className: consumerClassName,
    ...(0, import_merge_props.mergeProps)(behavioralTrackingProps, ariaProps)
  };
  const customTrigger = (0, import_react.useMemo)(
    () => getCustomTriggerSlot(remainingProps.children),
    [remainingProps.children]
  );
  const {
    targetRef,
    triggerRef,
    overlayRef,
    overlayState,
    overlayProps: _overlayProps,
    overlayTriggerProps,
    overlayContainerProps
  } = (0, import_useOverlayWithTrigger.useOverlayWithTrigger)({
    placement: "bottom-start",
    placementFallbacks: ["bottom-end", "top-start", "top-end"],
    offset: 4,
    shouldCloseOnBlur: false,
    onOpenChange: (isOpen2) => {
      if (!isOpen2 && !formFieldMessagesState.isTouched && import_SegmentsCache.segmentsClient.hasAccessRights !== false) {
        formFieldMessagesDispatch?.({
          type: import_formfield_messages_state_reducer.FormFieldMessagesAction.SETISTOUCHED,
          payload: { isTouched: true }
        });
      }
      if (!isOpen2) {
        (0, import_segmentValidationHelper.validateSegments)(
          segments,
          onChange,
          formFieldMessagesDispatch,
          formFieldMessagesState.formControlMessages,
          errorId,
          true
        ).catch(() => void 0);
      }
      onOpenChange?.(isOpen2);
    },
    additionalElevation: "overlay",
    disabled
  });
  const { trapFocus, ...overlayProps } = _overlayProps;
  const { onOpenTooltip, isOpen, textEllipsisRef } = (0, import_useTooltipHandler.useTooltipHandler)(
    !customTrigger && variant === "default"
  );
  const overlayAriaLabelledById = (0, import_useId.useId)();
  (0, import_react.useImperativeHandle)(
    forwardedRef,
    () => ({
      element: internalButtonRef.current,
      triggerRef: internalButtonRef.current,
      inputRef: inputRef.current,
      focus: () => internalButtonRef.current?.focus(),
      open: overlayState.open,
      close: overlayState.close,
      validate: async () => {
        formFieldMessagesDispatch?.({
          type: import_formfield_messages_state_reducer.FormFieldMessagesAction.SETISTOUCHED,
          payload: { isTouched: true }
        });
        if (import_SegmentsCache.segmentsClient.hasAccessRights === false) {
          formFieldMessagesDispatch?.({
            type: import_formfield_messages_state_reducer.FormFieldMessagesAction.ADDMESSAGE,
            payload: {
              id: errorId,
              variant: "error",
              validityStateKey: "customError",
              messageType: "formcontrol",
              showError: true,
              message: intl.formatMessage(
                import_segmentMessages.SEGMENTSELECTOR_MESSAGES.noReadAndWriteAccess
              )
            }
          });
        }
        (0, import_segmentValidationHelper.validateSegments)(
          segments,
          onChange,
          formFieldMessagesDispatch,
          formFieldMessagesState.formControlMessages,
          errorId,
          true
        ).catch(() => void 0);
      }
    }),
    [
      errorId,
      formFieldMessagesDispatch,
      formFieldMessagesState.formControlMessages,
      intl,
      onChange,
      overlayState.close,
      overlayState.open,
      segments
    ]
  );
  const buttonRef = (0, import_useMergeRefs.useMergeRefs)([triggerRef, targetRef, internalButtonRef]);
  (0, import_react.useEffect)(() => {
    (0, import_segmentValidationHelper.validateSegments)(
      segments,
      onChange,
      formFieldMessagesDispatch,
      formFieldMessagesState.formControlMessages,
      errorId
    ).catch(() => void 0);
  }, [
    errorId,
    formFieldMessagesDispatch,
    formFieldMessagesState.formControlMessages,
    onChange,
    segments,
    stringifiedValue
  ]);
  const providerValue = (0, import_react.useMemo)(
    () => ({
      isLoading,
      displayValue: text,
      triggerProps: {
        ...overlayTriggerProps,
        "data-dt-component": COMPONENT_NAME
      },
      buttonRef,
      showError: formFieldMessagesState.showError,
      ...ariaProps
    }),
    [
      isLoading,
      text,
      overlayTriggerProps,
      buttonRef,
      formFieldMessagesState.showError,
      ariaProps
    ]
  );
  const stratoLogger = (0, import_react.useMemo)(() => {
    return import_sdk_logger.createStratoLogger.bind({
      name,
      id: propId,
      ariaLabel: triggerAriaProps["aria-label"]
    })(COMPONENT_NAME);
  }, [name, propId, triggerAriaProps]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_SegmentsTriggerContext.SegmentsTriggerContext.Provider, { value: providerValue, children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_Tooltip.Tooltip,
        {
          text: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TooltipMessages.TooltipMessages, { ariaLabel: triggerAriaProps["aria-label"] }),
          className: styles.overwriteTooltipElementPadding,
          onOpenChange: onOpenTooltip,
          open: isOpen,
          children: customTrigger ?? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_SegmentSelectorTrigger.SegmentSelectorTrigger,
            {
              textEllipsisRef,
              ...segmentSelectorProps
            }
          )
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_VisuallyHidden.VisuallyHidden, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        "input",
        {
          ref: inputRef,
          id,
          value: segments.length > 0 ? stringifiedValue : "",
          onChange: () => void 0,
          "aria-hidden": true,
          readOnly,
          disabled,
          name,
          required,
          tabIndex: -1,
          "aria-disabled": ariaDisabled,
          onInvalid
        }
      ) })
    ] }),
    overlayState.isOpen && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_StratoLoggerContext.StratoLoggerContext.Provider, { value: stratoLogger, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_OverlayContainer.OverlayContainer,
      {
        "data-testid": dataTestId,
        ...overlayContainerProps,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_FocusScope.FocusScope, { autoFocus: true, contain: trapFocus, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          "div",
          {
            ref: overlayRef,
            className: import_Overlay_css.overlayCSS,
            ...overlayProps,
            "aria-labelledby": overlayAriaLabelledById,
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_SegmentsUIContext.UISegmentsProvider,
              {
                readOnly,
                overlayState,
                children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Overlay.SegmentsOverlay, { labelId: overlayAriaLabelledById })
              }
            )
          }
        ) })
      }
    ) })
  ] });
});
const SegmentSelectorRoot = (0, import_FormFieldMessageContextProvider.withFormFieldMessageContextProvider)(SegmentSelectorFwd);
const SegmentSelector = Object.assign(SegmentSelectorRoot, {
  CustomTrigger: import_CustomTrigger.CustomTrigger
});
SegmentSelector.displayName = COMPONENT_NAME;
