// @dynatrace/strato-components/charts/honeycomb/components/toolbar/HoneycombToolbar.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var HoneycombToolbar_exports = {};
__export(HoneycombToolbar_exports, {
  HoneycombToolbar: () => HoneycombToolbar
});
module.exports = __toCommonJS(HoneycombToolbar_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_HoneycombDownloadDataButton = require("./buttons/HoneycombDownloadDataButton.js");
var import_useBehavioralTrackingProps = require("../../../../core/hooks/useBehavioralTrackingProps.js");
var import_useTooltipReducer = require("../../../core/components/overlay-tooltip/useTooltipReducer.js");
var import_ChartToolbarRenderer = require("../../../core/components/toolbar/ChartToolbarRenderer.js");
var import_use_honeycomb_data = require("../../hooks/use-honeycomb-data.js");
var import_use_honeycomb_toolbar_config = require("../../hooks/use-honeycomb-toolbar-config.js");
var import_selectors = require("../../store/selectors.js");
var import_build_and_download_csv = require("../../utils/build-and-download-csv.js");
const HoneycombToolbar = () => {
  const toolbarVisible = (0, import_selectors.useHoneycombToolbarVisibleState)();
  const dispatch = (0, import_useTooltipReducer.useTooltipReducer)();
  const toolbarConfig = (0, import_use_honeycomb_toolbar_config.useHoneycombToolbarConfig)();
  const rawData = (0, import_use_honeycomb_data.useHoneycombData)();
  if ((0, import_lodash_es.isUndefined)(toolbarConfig)) {
    return null;
  }
  const isDownloadEnabled = toolbarConfig?.downloadCSV?.enabled ?? toolbarConfig?.downloadData?.enabled ?? false;
  const defaultToolbarPlacement = "top-right";
  const handleMouseEnter = () => {
    dispatch({ type: "RESET_TOOLTIP" })();
  };
  const handleDownloadDataToolbar = () => (0, import_build_and_download_csv.buildAndDownloadCsv)(rawData);
  const showDefaultToolbarDownloadButton = isDownloadEnabled && !toolbarConfig?.intents;
  const showMenuToolbarButton = isDownloadEnabled && !!toolbarConfig?.intents?.length;
  const initialA11yState = showDefaultToolbarDownloadButton ? [{ download: 0 }] : [{ intent: 0 }];
  const behaviorTrackingProps = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(
    toolbarConfig ?? {}
  )[0];
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ChartToolbarRenderer.ChartToolbarRenderer,
    {
      visible: toolbarVisible,
      collapsed: toolbarConfig.collapsed,
      draggable: true,
      placement: toolbarConfig.placement || defaultToolbarPlacement,
      initialA11yState,
      onMouseEnter: handleMouseEnter,
      menuDisplayMode: "never",
      isDownloadEnabled: showMenuToolbarButton,
      intents: toolbarConfig?.intents,
      handleDownloadDataToolbar,
      removeDragHandle: true,
      ...behaviorTrackingProps,
      children: showDefaultToolbarDownloadButton && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_HoneycombDownloadDataButton.HoneycombDownloadDataButton, {})
    }
  );
};
HoneycombToolbar["displayName"] = "HoneycombToolbar";
