// @dynatrace/strato-components/esm/tables/DataTable/features/RowOrder/row-order-model.js
import { isDev } from "../../../../core/utils/is-node-environment.js";
import { DATA_TABLE_ROW_EXPANDED_ID_SUFFIX } from "../../constants.js";
import { memo } from "../../tanstack-table-core-override/src/index.js";
function getRowOrderRowModel() {
  return (table) => {
    return memo(
      () => [
        table.getPreRowOrderRowModel(),
        table.options.enableRowOrdering,
        table.getState().rowOrder,
        table.options.state.sorting
      ],
      (rowModel) => {
        const rowOrder = table.getState().rowOrder;
        if (table.options.enableRowOrdering !== true || table.options.state.sorting && table.options.state.sorting.length !== 0) {
          return rowModel;
        }
        const { flatRows, rows, rowsById } = rowModel;
        if (rowOrder === void 0) {
          table.setRowOrder(flatRows.map((row) => row.id));
          return rowModel;
        }
        const orderRows = (rows2) => {
          const orderedData = [...rows2];
          orderedData.sort(
            (rowA, rowB) => rowOrderCompareFn(rowOrder, rowA, rowB)
          );
          orderedData.forEach((row) => {
            if (row.subRows?.length) {
              row.subRows = orderRows(row.subRows);
            }
          });
          return orderedData;
        };
        const newRowsById = Object.keys(rowsById).sort((a, b) => rowOrderCompareStrFn(rowOrder, a, b)).reduce((obj, key) => {
          obj[key] = rowsById[key];
          return obj;
        }, {});
        return {
          flatRows: orderRows(flatRows),
          rows: orderRows(rows),
          rowsById: newRowsById
        };
      },
      {
        key: isDev() && "_getRowOrderRowModel",
        debug: () => table.options.debugAll ?? table.options.debugRows
      }
    );
  };
}
function rowOrderCompareFn(rowOrder, a, b) {
  let indexA = rowOrder.indexOf(a.id);
  if (indexA === -1) {
    indexA = rowOrder.indexOf(`${a.id}${DATA_TABLE_ROW_EXPANDED_ID_SUFFIX}`);
  }
  let indexB = rowOrder.indexOf(b.id);
  if (indexA === -1) {
    indexB = rowOrder.indexOf(`${b.id}${DATA_TABLE_ROW_EXPANDED_ID_SUFFIX}`);
  }
  return indexA - indexB;
}
function rowOrderCompareStrFn(rowOrder, a, b) {
  let indexA = rowOrder.indexOf(a);
  if (indexA === -1) {
    indexA = rowOrder.indexOf(`${a}${DATA_TABLE_ROW_EXPANDED_ID_SUFFIX}`);
  }
  let indexB = rowOrder.indexOf(b);
  if (indexA === -1) {
    indexB = rowOrder.indexOf(`${b}${DATA_TABLE_ROW_EXPANDED_ID_SUFFIX}`);
  }
  return indexA - indexB;
}
export {
  getRowOrderRowModel
};
