// @dynatrace/strato-components/esm/tables/DataTable/features/RowOrder/RowOrder.js
import { jsx, jsxs } from "react/jsx-runtime";
import {
  DndContext as DragAndDropContext,
  DragOverlay,
  KeyboardSensor,
  MouseSensor,
  TouchSensor,
  useSensor,
  useSensors
} from "@dnd-kit/core";
import { useSortable } from "@dnd-kit/sortable";
import { functionalUpdate } from "@tanstack/react-table";
import clsx from "clsx";
import {
  useCallback,
  useMemo,
  useRef,
  useState
} from "react";
import { createPortal } from "react-dom";
import { useIntl } from "react-intl";
import { DragAllDirectionIcon } from "@dynatrace/strato-icons";
import { getRowOrderRowModel } from "./row-order-model.js";
import {
  createRestrictToTableRowAreaModifier,
  createRowOrderCollisionDetection,
  getAllRowsInTableBody,
  getDragHandleForRow,
  getDragIndexes,
  resetAllStylesForRows,
  transformRowOnDragOver
} from "./row-order-utils.js";
import * as styles from "./RowOrder.sty.js";
import { Button } from "../../../../buttons/button/Button.js";
import { useControllableState } from "../../../../core/hooks/useControllableState.js";
import { useModality } from "../../../../core/hooks/useModality.js";
import { focusRingStatic } from "../../../../core/styles/focusRingStatic.sty.js";
import { getProperInertValue } from "../../../../core/utils/get-inert-value.js";
import { devLog } from "../../../../core/utils/logging.js";
import { Grid } from "../../../../layouts/grid/Grid.js";
import { DataTableDefaultCell } from "../../components/DataTableDefaultCell.js";
import { DataTableRow } from "../../components/DataTableRow.js";
import {
  DATA_TABLE_BUILTIN_COLUMN_MIN_WIDTH,
  DATA_TABLE_ROW_DRAG_COLUMN_ID,
  DATA_TABLE_ROW_HEIGHT_MAP
} from "../../constants.js";
import { useDataTable } from "../../contexts/DataTableContext.js";
import { assignRowMethods } from "../../hooks/useTable/utils.js";
import { dataTableMessages } from "../../messages.js";
import { RowDetailsElement } from "../RowDetails/RowDetailsElement.js";
import * as thresholdStyles from "../Thresholds/Thresholds.sty.js";
const SCROLL_JUMP_THRESHOLD = DATA_TABLE_ROW_HEIGHT_MAP.default * 10;
function isRowOrderConfigObject(rowOrdering) {
  return typeof rowOrdering === "object";
}
function useRowOrder(props, options) {
  const { rowOrdering } = props;
  const disableRowDragAndDropFn = typeof rowOrdering === "object" ? rowOrdering.disableRowDragAndDrop : void 0;
  const enableRowDragAndDropCallback = useCallback(
    (row) => {
      if (!disableRowDragAndDropFn) {
        return true;
      }
      return !disableRowDragAndDropFn(row.original);
    },
    [disableRowDragAndDropFn]
  );
  const [rowOrder, onRowOrderChange] = useControllableState({
    value: props.rowOrder,
    defaultValue: props.defaultRowOrder,
    onChange: props.onRowOrderChange
  });
  options.enableRowOrderDragAndDrop = false;
  if (rowOrdering === false || rowOrdering === void 0) {
    options.enableRowOrdering = false;
    return;
  }
  options.enableRowOrdering = true;
  options.getRowOrderRowModel = getRowOrderRowModel();
  options.onRowOrderChange = onRowOrderChange;
  options.state = { ...options.state, rowOrder };
  if (isRowOrderConfigObject(rowOrdering)) {
    if (rowOrdering.enableDragAndDrop === true && rowOrdering.disableRowDragAndDrop !== void 0) {
      options.lockDisabledRows = rowOrdering.lockDisabledRows;
      options.enableRowOrderDragAndDrop = enableRowDragAndDropCallback;
      options.autoResetPageIndex = false;
    } else {
      options.enableRowOrderDragAndDrop = rowOrdering.enableDragAndDrop === true;
    }
  }
}
const RowDragHandleCell = ({
  rowId,
  disabled,
  isDraggingActive
}) => {
  const intl = useIntl();
  const modality = useModality();
  const { attributes, listeners } = useSortable({
    id: rowId
  });
  const isKeyboardDragging = isDraggingActive && modality === "keyboard";
  const rowDragHandleLabel = intl.formatMessage(
    dataTableMessages.rowDragAndDropLabel,
    { rowId }
  );
  return /* @__PURE__ */ jsx(DataTableDefaultCell, { children: /* @__PURE__ */ jsx(
    Button,
    {
      size: "condensed",
      variant: "default",
      ...attributes,
      ...!disabled ? listeners : void 0,
      "aria-disabled": disabled,
      "aria-label": rowDragHandleLabel,
      className: clsx(
        styles.dndHandle({
          disabled
        }),
        // Apply the focus ring unconditionally (when: 'always')
        // to the dragged row's handle during keyboard dragging.
        isKeyboardDragging && focusRingStatic({
          minimal: false,
          color: "neutral",
          when: "always"
        })
      ),
      "data-dt-name": "RowOrderDragAndDrop",
      children: /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(DragAllDirectionIcon, {}) })
    }
  ) });
};
const ROW_DRAG_COLUMN = {
  id: DATA_TABLE_ROW_DRAG_COLUMN_ID,
  alignment: "center",
  accessorFn: () => "",
  header: () => {
    return "";
  },
  width: "content",
  minWidth: DATA_TABLE_BUILTIN_COLUMN_MIN_WIDTH,
  enableSorting: false,
  enableResizing: false,
  enableUserActions: false,
  enableHiding: false,
  meta: {
    builtin: true,
    hasCustomCellRenderer: true,
    hasCustomHeaderRenderer: true,
    builtinPlacement: "prefix",
    cellDeps: (cell) => {
      return [cell.row.getCanDrag()];
    }
  },
  cell: ({ row, table }) => /* @__PURE__ */ jsx(
    RowDragHandleCell,
    {
      rowId: row.id,
      disabled: !row.getCanDrag(),
      isDraggingActive: table.getState().draggingRowInfo?.id === row.id
    }
  )
};
const DataTableRowOrder = {
  createTable(table) {
    table.getPreRowOrderRowModel = () => table.getExpandedRowModel();
    table.setDraggingRowInfo = (draggingRowInfo) => {
      table.setState((state) => ({
        ...state,
        draggingRowInfo
      }));
    };
    table.getRowOrderRowModel = () => {
      if (!table._getRowOrderRowModel && table.options.getRowOrderRowModel) {
        table._getRowOrderRowModel = table.options.getRowOrderRowModel(table);
      }
      if (!table._getRowOrderRowModel) {
        return table.getPreRowOrderRowModel();
      }
      return table._getRowOrderRowModel();
    };
    table.setRowOrder = (updater) => {
      const newValue = functionalUpdate(
        updater,
        table.getState().rowOrder ?? []
      );
      table._queue(() => {
        table.options.onRowOrderChange?.(newValue);
      });
    };
    assignRowMethods(table, {
      getCanDrag() {
        const { enableRowOrdering, enableRowOrderDragAndDrop } = table.options;
        const sortingState = table.getState().sorting;
        if (enableRowOrdering !== true || // always disable the drag and drop handle if the rows are sorted
        sortingState !== void 0 && sortingState.length > 0) {
          return false;
        }
        if (enableRowOrderDragAndDrop !== void 0 && typeof enableRowOrderDragAndDrop === "function") {
          return enableRowOrderDragAndDrop(this);
        }
        return true;
      },
      disabledRowDragAndDropLockPosition: table.options.lockDisabledRows ?? false
    });
  }
};
function RowOrderingDragAndDropContext({
  children,
  rows,
  virtualColumns,
  rowSeparation
}) {
  const table = useDataTable();
  const modality = useModality();
  const tableHeaderHeight = table.getTotalHeaderHeight();
  const sensors = useSensors(
    useSensor(MouseSensor, {}),
    useSensor(TouchSensor, {}),
    useSensor(KeyboardSensor, {})
  );
  const [draggingElement, setDraggingElement] = useState(
    void 0
  );
  const containedVariant = table.getVisualVariantOptions().contained;
  const virtualizationContainer = table.getRowVirtualizationContainer();
  const draggingRowInfo = table.getState().draggingRowInfo;
  const tableGridElement = virtualizationContainer?.querySelector("[role=grid]");
  const overlayContainer = useRef(null);
  const scrollHandlerRef = useRef(null);
  const rowOrderCollisionDetection = useMemo(
    () => createRowOrderCollisionDetection(draggingRowInfo),
    [draggingRowInfo]
  );
  const restrictToTableRowArea = useMemo(
    () => createRestrictToTableRowAreaModifier(
      tableHeaderHeight,
      containedVariant,
      virtualizationContainer
    ),
    [tableHeaderHeight, virtualizationContainer, containedVariant]
  );
  const rowOrderWhileDragging = useRef([]);
  if (!tableGridElement || !virtualizationContainer) {
    return;
  }
  const overlayWidth = tableGridElement.parentElement?.clientWidth ?? 0;
  const originalRowOrder = table.getState().rowOrder ?? rows.map((row) => row.id);
  const resetDragging = () => {
    if (virtualizationContainer && scrollHandlerRef.current) {
      virtualizationContainer.removeEventListener(
        "scroll",
        scrollHandlerRef.current
      );
      scrollHandlerRef.current = null;
    }
    table.setDraggingRowInfo(void 0);
    setDraggingElement(void 0);
    resetAllStylesForRows(getAllRowsInTableBody(tableGridElement));
  };
  const handleOnDragStart = ({ active }) => {
    if (active.data.current) {
      const element = createDraggingElement(
        overlayContainer,
        active,
        rows,
        virtualColumns,
        {
          width: overlayWidth,
          gridTemplateColumns: table.getGridTemplateColumns(),
          rowSeparation,
          hasThresholds: table.options.hasThresholds,
          rowThresholds: table.options.rowThresholds,
          rowDetailsTemplate: table.options.rowDetailsTemplate
        }
      );
      setDraggingElement(element);
      requestAnimationFrame(() => {
        if (overlayContainer.current?.parentElement) {
          overlayContainer.current.parentElement.scrollLeft = tableGridElement?.parentElement?.scrollLeft ?? 0;
        }
      });
      table.setDraggingRowInfo({
        id: active.id,
        index: active.data.current.rowIndex,
        row: active.data.current.row
      });
      rowOrderWhileDragging.current = originalRowOrder ? [...originalRowOrder] : [];
      let lastScrollOffset = virtualizationContainer?.scrollTop ?? 0;
      const handleScroll = () => {
        if (!virtualizationContainer) {
          return;
        }
        const currentOffset = virtualizationContainer.scrollTop;
        const scrollDelta = Math.abs(currentOffset - lastScrollOffset);
        if (scrollDelta >= SCROLL_JUMP_THRESHOLD) {
          virtualizationContainer.scrollTop = lastScrollOffset;
          return;
        }
        lastScrollOffset = currentOffset;
      };
      scrollHandlerRef.current = handleScroll;
      virtualizationContainer?.addEventListener("scroll", handleScroll);
    }
  };
  const handleOnDragEnd = ({
    active,
    over
  }) => {
    if (over === void 0 || over === null) {
      devLog("[Table] On drag end: over info is undefined or null", over);
      resetDragging();
      return;
    }
    if (draggingRowInfo === void 0) {
      devLog(
        "[Table] On drag end: dragging row info is undefined",
        draggingRowInfo
      );
      resetDragging();
      return;
    }
    if (over.id === active.id) {
      resetDragging();
      return;
    }
    const { fromIndex, toIndex } = getDragIndexes(
      originalRowOrder,
      active.id,
      // for us the IDs are always strings
      over.id
      // for us the IDs are always strings
    );
    const newRowOrder = [...originalRowOrder];
    const activeRowSubRowsCount = draggingRowInfo.row.getLeafRows().length;
    const overRowSubRowsCount = (over.data.current?.row).getLeafRows().length;
    const rowsToMove = newRowOrder.splice(fromIndex, 1 + activeRowSubRowsCount);
    newRowOrder.splice(
      fromIndex < toIndex ? toIndex + overRowSubRowsCount - activeRowSubRowsCount : toIndex,
      0,
      ...rowsToMove
    );
    table.setRowOrder(newRowOrder);
    resetDragging();
    virtualizationContainer?.scrollBy({ top: 1, behavior: "smooth" });
    virtualizationContainer?.scrollBy({ top: -1, behavior: "smooth" });
    requestAnimationFrame(() => {
      const dragHandle = getDragHandleForRow(tableGridElement, toIndex);
      dragHandle?.focus(
        modality !== "keyboard" ? { focusVisible: false } : void 0
      );
    });
  };
  return /* @__PURE__ */ jsxs(
    DragAndDropContext,
    {
      sensors,
      collisionDetection: rowOrderCollisionDetection,
      modifiers: [restrictToTableRowArea],
      onDragStart: handleOnDragStart,
      onDragEnd: handleOnDragEnd,
      onDragOver: (event) => transformRowOnDragOver(
        tableGridElement,
        event,
        originalRowOrder,
        table.getVisualVariantOptions().rowSeparation === "none",
        draggingRowInfo?.row
      ),
      onDragCancel: () => resetDragging(),
      children: [
        children,
        // We need to portal the drag and drop overlay to escape CSS transforms on ancestor elements.
        // This ensures transforms on parent containers don't affect the position of the drag overlay.
        createPortal(
          /* @__PURE__ */ jsx(
            DragOverlay,
            {
              dropAnimation: null,
              className: styles.rowDndOverlay,
              style: { width: overlayWidth, height: "max-content" },
              children: draggingElement
            }
          ),
          document.body
        )
      ]
    }
  );
}
function createDraggingElement(overlayContainer, active, rows, virtualColumns, rowConfig) {
  const activeId = active.id;
  const draggingOriginalRow = active.data.current?.row;
  const draggingVirtualRow = active.data.current?.virtualRow;
  const {
    gridTemplateColumns,
    rowSeparation,
    hasThresholds,
    rowThresholds,
    rowDetailsTemplate
  } = rowConfig;
  if (draggingOriginalRow === void 0 || draggingVirtualRow === void 0) {
    devLog(
      `[DataTable] Dragging row or virtual row with row id ${activeId} is undefined.`
    );
    return;
  }
  const isRowDetailsOpen = draggingOriginalRow.getIsDetailsOpen();
  const draggingVirtualDetailsRow = {
    ...draggingVirtualRow,
    index: draggingVirtualRow.index + 1
  };
  return /* @__PURE__ */ jsxs(
    Grid,
    {
      ref: overlayContainer,
      gridTemplateColumns,
      padding: 0,
      minWidth: "max-content",
      gap: 0,
      ...{ inert: getProperInertValue(true) },
      className: clsx({
        [thresholdStyles.hasPillThresholds]: hasThresholds && rowThresholds?.some((threshold) => threshold.type === "pill")
      }),
      style: {
        position: "relative"
      },
      children: [
        /* @__PURE__ */ jsx(
          DataTableRow,
          {
            row: draggingOriginalRow,
            virtualRow: draggingVirtualRow,
            virtualItemIndex: 0,
            rowCount: rows.length,
            virtualColumns,
            isDraggedRow: false,
            isInDragOverlay: true,
            rowParity: "even"
          }
        ),
        isRowDetailsOpen ? /* @__PURE__ */ jsx(
          RowDetailsElement,
          {
            originalRow: draggingOriginalRow,
            virtualRow: draggingVirtualDetailsRow,
            virtualItemIndex: 1,
            rowSeparation,
            rowParity: "even",
            children: rowDetailsTemplate?.({
              row: draggingOriginalRow.original
            })
          }
        ) : null
      ]
    }
  );
}
export {
  DataTableRowOrder,
  ROW_DRAG_COLUMN,
  RowOrderingDragAndDropContext,
  useRowOrder
};
