// @dynatrace/strato-components/esm/charts/core/components/axes/components/CategoryLeftAxis.js
import { jsx } from "react/jsx-runtime";
import { getTicks } from "@visx/scale";
import { DEFAULT_TRUNCATION_MODE } from "../../../constants.js";
import { AxisLeft } from "../AxisLeft.js";
import { axisFontProps, EMOJI_REGEX } from "../constants.js";
import { DivAxisWrapper } from "./DivAxisWrapper.js";
import { TickLabel } from "./TickLabel.js";
import { getYTickMaxSize } from "../utils/calculate-axis-size.js";
import { getTickLabelPropsForCategoryAxis } from "../utils/get-tick-label-props.js";
import { isTickLabelSpaceAvailable } from "../utils/is-tick-label-visible.js";
import { getCategoryLeftAxisProps } from "./utils/getCategoryLeftAxisProps.js";
const TICK_LABEL_PROPS = getTickLabelPropsForCategoryAxis("left");
const CategoryLeftAxis = (props) => {
  const {
    scale,
    containerWidth,
    containerHeight,
    label,
    categories,
    truncationMode = DEFAULT_TRUNCATION_MODE,
    isVerticalLayout = true
  } = props;
  const ticks = getTicks(scale);
  const tickNames = ticks.map((tickId) => categories[tickId]) ?? "Unknown";
  const maxSize = getYTickMaxSize(containerWidth, label);
  const tickFormat = (_, index) => tickNames[index];
  const isTickLabelVisible = isTickLabelSpaceAvailable(scale.step());
  const tickNameHasEmoji = tickNames.some((name) => EMOJI_REGEX.test(name));
  return isTickLabelVisible ? /* @__PURE__ */ jsx(
    DivAxisWrapper,
    {
      divAxisStyle: axisFontProps,
      ticks: tickNames,
      label,
      children: /* @__PURE__ */ jsx(
        AxisLeft,
        {
          role: "graphics-symbol",
          "aria-label": "category axis left",
          scale,
          containerSize: containerHeight,
          width: containerWidth,
          tickComponent: ({
            x: renderX,
            y: renderY,
            ...rest
          }) => {
            const {
              x,
              y,
              foreignObjectInlineStyle,
              foreignObjectContentInlineStyle
            } = getCategoryLeftAxisProps(
              renderX,
              renderY,
              isVerticalLayout,
              maxSize,
              tickNameHasEmoji
            );
            return /* @__PURE__ */ jsx(
              TickLabel,
              {
                foreignObjectInlineStyle,
                foreignObjectContentInlineStyle,
                truncationMode,
                maxSize,
                x,
                y,
                ...rest
              }
            );
          },
          tickValues: ticks,
          tickLabelProps: TICK_LABEL_PROPS,
          tickFormat,
          label
        }
      )
    }
  ) : null;
};
CategoryLeftAxis["displayName"] = "CategoryLeftAxis";
export {
  CategoryLeftAxis
};
