// @dynatrace/strato-components/esm/charts/core/components/axes/components/CategoryBottomAxis.js
import { jsx } from "react/jsx-runtime";
import { getTicks } from "@visx/scale";
import { DivAxisWrapper } from "./DivAxisWrapper.js";
import { AxisBottom } from "../AxisBottom.js";
import { axisFontProps, EMOJI_REGEX } from "../constants.js";
import { TickLabel } from "./TickLabel.js";
import { getXTickMaxSize } from "../utils/calculate-axis-size.js";
import { getTickLabelPropsForCategoryAxis } from "../utils/get-tick-label-props.js";
import { isTickLabelSpaceAvailable } from "../utils/is-tick-label-visible.js";
import { getCategoryBottomAxisProps } from "./utils/getCategoryBottomAxisProps.js";
const CategoryBottomAxis = (props) => {
  const {
    scale,
    containerWidth,
    categories,
    containerHeight,
    tickLabelLayout = "horizontal",
    label,
    truncationMode = "end"
  } = props;
  const ticks = getTicks(scale);
  const tickNames = ticks.map((tickId) => categories[tickId]) ?? "Unknown";
  const isTickRotated = tickLabelLayout === "vertical";
  const isTickLabelVisible = isTickLabelSpaceAvailable(scale.step());
  const maxSize = getXTickMaxSize(
    containerHeight,
    containerWidth,
    isTickRotated,
    scale,
    label
  );
  const tickLabelProps = getTickLabelPropsForCategoryAxis(
    "bottom",
    tickLabelLayout
  );
  const tickNameHasEmoji = tickNames.some((name) => EMOJI_REGEX.test(name));
  const tickFormat = (_, index) => tickNames[index];
  return isTickLabelVisible ? /* @__PURE__ */ jsx(
    DivAxisWrapper,
    {
      divAxisStyle: axisFontProps,
      ticks: tickNames,
      label,
      isTickVertical: true,
      isTickRotated,
      children: /* @__PURE__ */ jsx(
        AxisBottom,
        {
          role: "graphics-symbol",
          "aria-label": "category axis bottom",
          scale,
          containerSize: containerWidth,
          height: containerHeight,
          tickComponent: ({
            x: renderX,
            y: renderY,
            ...rest
          }) => {
            const {
              x,
              y,
              foreignObjectInlineStyle,
              foreignObjectContentInlineStyle
            } = getCategoryBottomAxisProps(
              renderX,
              isTickRotated,
              maxSize,
              tickNameHasEmoji
            );
            return /* @__PURE__ */ jsx(
              TickLabel,
              {
                foreignObjectInlineStyle,
                foreignObjectContentInlineStyle,
                isTickRotated,
                truncationMode,
                maxSize,
                x,
                y,
                ...rest
              }
            );
          },
          tickValues: ticks,
          tickLabelProps,
          tickFormat,
          label,
          width: containerWidth
        }
      )
    }
  ) : null;
};
CategoryBottomAxis["displayName"] = "CategoryBottomAxis";
export {
  CategoryBottomAxis
};
