// @dynatrace/strato-components/esm/charts/core/components/axes/components/DivAxisWrapper.js
import { jsx, jsxs } from "react/jsx-runtime";
import { Flex } from "../../../../../layouts/flex/Flex.js";
import {
  AXIS_LABEL_HEIGHT,
  X_AXES_LABEL_OFFSET,
  X_AXES_TICK_LINE_OFFSET,
  Y_AXES_TICK_LINE_OFFSET,
  Y_AXIS_LABEL_OFFSET
} from "../constants.js";
const TICK_MIN_WIDTH = 8;
const TICK_MIN_HEIGHT = 12;
const getDivContainerAxisStyle = (isTickVertical, isTickRotated) => {
  let tickStyle;
  let labelStyle;
  const rotatedTextStyle = {
    writingMode: "vertical-rl"
  };
  if (isTickVertical) {
    tickStyle = {
      marginTop: X_AXES_TICK_LINE_OFFSET,
      minHeight: isTickRotated ? TICK_MIN_HEIGHT : 0,
      whiteSpace: "nowrap",
      overflow: "hidden",
      ...isTickRotated ? rotatedTextStyle : {}
    };
    labelStyle = {
      marginTop: X_AXES_LABEL_OFFSET,
      whiteSpace: "nowrap",
      overflow: "hidden"
    };
  } else {
    tickStyle = {
      marginRight: Y_AXES_TICK_LINE_OFFSET,
      minWidth: TICK_MIN_WIDTH,
      whiteSpace: "nowrap"
    };
    labelStyle = {
      marginRight: Y_AXIS_LABEL_OFFSET,
      minWidth: AXIS_LABEL_HEIGHT,
      whiteSpace: "nowrap",
      ...rotatedTextStyle
    };
  }
  return { tickStyle, labelStyle };
};
const DivAxisWrapper = ({
  ticks,
  label,
  isTickVertical = false,
  isTickRotated = false,
  divAxisStyle = {},
  children
}) => {
  const { tickStyle, labelStyle } = getDivContainerAxisStyle(
    isTickVertical,
    isTickRotated
  );
  const containerAxisStyle = {
    position: "relative",
    /* A combination of pointer Events is needed here and in visibleSVGAxisStyle to allow the correct functioning with the resizer in the bottom legend when the tick labels are rotated and truncated (edge case) */
    pointerEvents: "none"
  };
  const visibleSvgAxisStyle = {
    position: "absolute",
    pointerEvents: "visible"
  };
  const hiddenDivAxisStyle = {
    visibility: "hidden",
    ...divAxisStyle
  };
  return /* @__PURE__ */ jsxs("div", { style: containerAxisStyle, children: [
    /* @__PURE__ */ jsx("div", { style: visibleSvgAxisStyle, children }),
    /* @__PURE__ */ jsxs(
      Flex,
      {
        style: hiddenDivAxisStyle,
        flexDirection: isTickVertical ? "column-reverse" : "row",
        gap: 0,
        children: [
          label && /* @__PURE__ */ jsx("span", { style: labelStyle, children: label }),
          /* @__PURE__ */ jsx(Flex, { flexDirection: isTickVertical ? "row" : "column", children: ticks?.map((tick, index) => /* @__PURE__ */ jsx("span", { style: tickStyle, children: tick }, `${index}_${tick}`)) })
        ]
      }
    )
  ] });
};
DivAxisWrapper["displayName"] = "DivAxisWrapper";
export {
  DivAxisWrapper
};
