// @dynatrace/strato-components/esm/charts/core/components/axes/components/TickLabel.js
import { jsx, jsxs } from "react/jsx-runtime";
import { Tooltip } from "../../../../..//overlays/tooltip/Tooltip.js";
import { isSafari } from "../../../utils/is-browser-safari.js";
import { charsFontWidth12pixelsSize400Map } from "../../text-ellipsis/chars-font-width-map.js";
import { truncateString } from "../../text-ellipsis/truncate-string.js";
const TickLabel = (props) => {
  const {
    foreignObjectInlineStyle,
    foreignObjectContentInlineStyle,
    formattedValue,
    maxSize,
    truncationMode,
    x,
    y
  } = props;
  if (!formattedValue) {
    return null;
  }
  const truncatedValue = truncateString(
    formattedValue,
    maxSize,
    charsFontWidth12pixelsSize400Map,
    truncationMode
  );
  const disableTooltip = formattedValue === truncatedValue;
  return /* @__PURE__ */ jsx("g", { children: /* @__PURE__ */ jsxs(
    "foreignObject",
    {
      x,
      y,
      height: foreignObjectInlineStyle.height,
      width: foreignObjectInlineStyle.width,
      style: { ...foreignObjectInlineStyle },
      children: [
        isSafari() && /* @__PURE__ */ jsx(
          "div",
          {
            style: { ...foreignObjectContentInlineStyle },
            title: formattedValue,
            children: truncatedValue
          }
        ),
        !isSafari() && /* @__PURE__ */ jsx(
          Tooltip,
          {
            disabled: disableTooltip,
            text: formattedValue,
            style: { display: "inline-block" },
            children: /* @__PURE__ */ jsx(
              "div",
              {
                style: {
                  ...foreignObjectContentInlineStyle,
                  position: "relative"
                },
                children: truncatedValue
              }
            )
          }
        )
      ]
    }
  ) });
};
export {
  TickLabel
};
