// @dynatrace/strato-components/esm/charts/core/components/axes/components/TimeBottomAxis.js
import { jsx } from "react/jsx-runtime";
import { AxisBottom } from "../AxisBottom.js";
import { axisFontProps } from "../constants.js";
import { DivAxisWrapper } from "./DivAxisWrapper.js";
import {
  bottomPanTickLabelDateProps,
  bottomTickLabelDateProps
} from "../utils/get-tick-label-props.js";
const TimeBottomAxis = (props) => {
  const {
    scale,
    containerWidth,
    containerHeight,
    label,
    stroke,
    ticks,
    tickColor,
    formattedTicks,
    shrinkTickEdges
  } = props;
  const tickFormat = (_, index) => formattedTicks[index];
  return /* @__PURE__ */ jsx(
    DivAxisWrapper,
    {
      divAxisStyle: axisFontProps,
      ticks: formattedTicks,
      label,
      isTickVertical: true,
      children: /* @__PURE__ */ jsx(
        AxisBottom,
        {
          role: "graphics-symbol",
          "aria-label": "time axis bottom",
          scale,
          containerSize: containerWidth,
          height: containerHeight,
          tickValues: ticks,
          tickLabelProps: shrinkTickEdges ? bottomPanTickLabelDateProps : bottomTickLabelDateProps,
          tickFormat,
          tickColor,
          label,
          stroke
        }
      )
    }
  );
};
TimeBottomAxis["displayName"] = "TimeBottomAxis";
export {
  TimeBottomAxis
};
