// @dynatrace/strato-components/esm/charts/core/components/axes/components/TimeTableHeaderAxis.js
import { jsx } from "react/jsx-runtime";
import { getTicks } from "@visx/scale";
import { DivAxisWrapper } from "./DivAxisWrapper.js";
import { formatDateTicks } from "../../../utils/format-date-ticks.js";
import { getTickValues } from "../../../utils/format-value.js";
import { getGanttHorizontalTimeAxisNumTicks } from "../../../utils/ticks.js";
import { AxisTableHeader } from "../AxisTableHeader.js";
import { axisFontProps } from "../constants.js";
import { getTickLabelPropsForGanttAxis } from "../utils/get-tick-label-props.js";
const TimeTableHeaderAxis = (props) => {
  const { scale, containerWidth, containerHeight, stroke, min, max } = props;
  const numTicks = getGanttHorizontalTimeAxisNumTicks(containerWidth);
  const ticks = getTicks(scale, numTicks);
  const tickValues = getTickValues(
    ticks,
    formatDateTicks
  );
  const tickFormat = (_, index) => tickValues[index];
  return /* @__PURE__ */ jsx(
    DivAxisWrapper,
    {
      divAxisStyle: axisFontProps,
      ticks: tickValues,
      isTickVertical: true,
      children: /* @__PURE__ */ jsx(
        AxisTableHeader,
        {
          role: "graphics-symbol",
          "aria-label": "time axis tableheader",
          scale,
          containerSize: containerWidth,
          height: containerHeight,
          numTicks,
          tickValues: ticks,
          tickLabelProps: getTickLabelPropsForGanttAxis(min, max),
          tickFormat,
          stroke
        }
      )
    }
  );
};
TimeTableHeaderAxis["displayName"] = "TimeTableHeaderAxis";
export {
  TimeTableHeaderAxis
};
