// @dynatrace/strato-components/esm/charts/core/components/axes/components/ValueBottomAxis.js
import { jsx } from "react/jsx-runtime";
import { DivAxisWrapper } from "./DivAxisWrapper.js";
import { AxisBottom } from "../AxisBottom.js";
import { axisFontProps } from "../constants.js";
import { getTickLabelPropsForValueAxis } from "../utils/get-tick-label-props.js";
const TICK_LABEL_PROPS = getTickLabelPropsForValueAxis("bottom");
const ValueBottomAxis = (props) => {
  const {
    scale,
    containerHeight,
    containerWidth,
    label,
    stroke,
    ticks,
    formattedTicks
  } = props;
  const tickFormat = (_, index) => formattedTicks[index];
  return /* @__PURE__ */ jsx(
    DivAxisWrapper,
    {
      divAxisStyle: axisFontProps,
      ticks: formattedTicks,
      label,
      isTickVertical: true,
      children: /* @__PURE__ */ jsx(
        AxisBottom,
        {
          role: "graphics-symbol",
          "aria-label": "value axis bottom",
          scale,
          containerSize: containerWidth,
          height: containerHeight,
          tickValues: ticks,
          tickLabelProps: TICK_LABEL_PROPS,
          tickFormat,
          label,
          stroke
        }
      )
    }
  );
};
ValueBottomAxis["displayName"] = "ValueBottomAxis";
export {
  ValueBottomAxis
};
