// @dynatrace/strato-components/esm/charts/core/components/axes/components/ValueRightAxis.js
import { jsx } from "react/jsx-runtime";
import { DivAxisWrapper } from "./DivAxisWrapper.js";
import { AxisRight } from "../AxisRight.js";
import { axisFontProps } from "../constants.js";
import { getTickLabelPropsForValueAxis } from "../utils/get-tick-label-props.js";
const TICK_LABEL_PROPS = getTickLabelPropsForValueAxis("right");
const ValueRightAxis = (props) => {
  const {
    scale,
    containerHeight,
    containerWidth,
    label,
    stroke,
    ticks,
    formattedTicks
  } = props;
  const tickFormat = (_, index) => formattedTicks[index];
  return /* @__PURE__ */ jsx(
    DivAxisWrapper,
    {
      divAxisStyle: axisFontProps,
      ticks: formattedTicks,
      label,
      children: /* @__PURE__ */ jsx(
        AxisRight,
        {
          role: "graphics-symbol",
          "aria-label": "value axis right",
          scale,
          containerSize: containerHeight,
          width: containerWidth,
          tickValues: ticks,
          tickLabelProps: TICK_LABEL_PROPS,
          tickFormat,
          label,
          stroke
        }
      )
    }
  );
};
ValueRightAxis["displayName"] = "ValueRightAxis";
export {
  ValueRightAxis
};
