// @dynatrace/strato-components/esm/charts/core/components/axes/components/ValueTableHeaderAxis.js
import { jsx } from "react/jsx-runtime";
import { getTicks } from "@visx/scale";
import { DivAxisWrapper } from "./DivAxisWrapper.js";
import { getTickValues } from "../../../utils/format-value.js";
import { getHorizontalNumTicks } from "../../../utils/ticks.js";
import { AxisTableHeader } from "../AxisTableHeader.js";
import { axisFontProps } from "../constants.js";
import { getTickLabelPropsForGanttAxis } from "../utils/get-tick-label-props.js";
const ValueTableHeaderAxis = (props) => {
  const {
    scale,
    containerWidth,
    containerHeight,
    formatter,
    stroke,
    min,
    max
  } = props;
  const numTicks = getHorizontalNumTicks(containerWidth);
  const ticks = getTicks(scale, numTicks);
  const tickValues = getTickValues(ticks, formatter);
  const tickFormat = (_, index) => tickValues[index];
  return /* @__PURE__ */ jsx(
    DivAxisWrapper,
    {
      divAxisStyle: axisFontProps,
      ticks: tickValues,
      isTickVertical: true,
      children: /* @__PURE__ */ jsx(
        AxisTableHeader,
        {
          role: "graphics-symbol",
          "aria-label": "value axis tableheader",
          scale,
          containerSize: containerWidth,
          height: containerHeight,
          numTicks,
          tickValues: ticks,
          tickLabelProps: getTickLabelPropsForGanttAxis(min, max),
          tickFormat,
          stroke
        }
      )
    }
  );
};
ValueTableHeaderAxis["displayName"] = "ValueTableHeaderAxis";
export {
  ValueTableHeaderAxis
};
