// @dynatrace/strato-components/esm/layouts/page-layout/PageLayout.js
import { jsx, jsxs } from "react/jsx-runtime";
import { clsx } from "clsx";
import {
  forwardRef,
  memo,
  useMemo,
  useSyncExternalStore
} from "react";
import { DetailsLayout } from "./details-layout/DetailsLayout.js";
import * as styles from "./PageLayout.sty.js";
import {
  PageLayoutPortalContext,
  usePageLayoutPortalContext
} from "./shared/PageLayoutPortalContext.js";
import {
  PageLayoutContent,
  PageLayoutDetailsWithControlBar,
  PageLayoutHeader,
  PageLayoutSidebar
} from "./shared/portal-slots.js";
import { contentDefaults } from "./shared/slot-defaults.js";
import { createSlotRegistry } from "./shared/slot-registry.js";
import { TeleportTarget } from "./shared/TeleportTarget.js";
import { useStableHost } from "./shared/useStableHost.js";
import { SidebarLayout } from "./sidebar-layout/SidebarLayout.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
function resolveOptionalSlot(config, target) {
  if (!config) {
    return null;
  }
  return { ...config, children: target };
}
function MainShell() {
  const { hosts, registry } = usePageLayoutPortalContext();
  const snapshot = useSyncExternalStore(
    registry.subscribe,
    registry.getSnapshot
  );
  const sidebarConfig = snapshot.sidebar;
  const detailsConfig = snapshot.details;
  const detailsLayoutContext = snapshot.detailsLayoutContext;
  const showTopLeftRadius = sidebarConfig !== null;
  const showTopRightRadius = detailsConfig !== null && detailsLayoutContext?.mode !== "collapsed";
  const contentClassName = clsx(
    styles.content,
    showTopLeftRadius && styles.contentTopLeftRadius,
    showTopRightRadius && styles.contentTopRightRadius
  );
  const sidebarTarget = useMemo(
    () => /* @__PURE__ */ jsx(TeleportTarget, { host: hosts.sidebar }),
    [hosts.sidebar]
  );
  const detailsTarget = useMemo(
    () => /* @__PURE__ */ jsx(TeleportTarget, { host: hosts.details }),
    [hosts.details]
  );
  const contentTarget = useMemo(
    () => /* @__PURE__ */ jsx(TeleportTarget, { host: hosts.content }),
    [hosts.content]
  );
  const sidebarSlot = useMemo(
    () => resolveOptionalSlot(snapshot.sidebar, sidebarTarget),
    [snapshot.sidebar, sidebarTarget]
  );
  const detailsSlot = useMemo(
    () => resolveOptionalSlot(snapshot.details, detailsTarget),
    [snapshot.details, detailsTarget]
  );
  const contentSlot = useMemo(() => {
    if (snapshot.content) {
      const { config, domProps, ref } = snapshot.content;
      return {
        children: contentTarget,
        config,
        domProps: {
          ...domProps,
          className: clsx(contentClassName, domProps.className)
        },
        ref
      };
    }
    return {
      children: void 0,
      config: { minWidth: contentDefaults.minWidth },
      domProps: { className: contentClassName }
    };
  }, [snapshot.content, contentTarget, contentClassName]);
  const sidebarContentSlot = useMemo(
    () => ({
      children: /* @__PURE__ */ jsx(DetailsLayout, { details: detailsSlot, content: contentSlot }),
      config: { minWidth: contentDefaults.minWidth },
      domProps: {}
    }),
    [detailsSlot, contentSlot]
  );
  return /* @__PURE__ */ jsx(SidebarLayout, { sidebar: sidebarSlot, content: sidebarContentSlot });
}
function HeaderShell() {
  const { hosts, registry } = usePageLayoutPortalContext();
  const snapshot = useSyncExternalStore(
    registry.subscribe,
    registry.getSnapshot
  );
  const headerConfig = snapshot.header;
  if (headerConfig === null) {
    return null;
  }
  return /* @__PURE__ */ jsx(
    "div",
    {
      ...headerConfig.domProps,
      ref: headerConfig.ref,
      className: clsx(styles.headerContainer, headerConfig.domProps.className),
      children: /* @__PURE__ */ jsx(TeleportTarget, { host: hosts.header })
    }
  );
}
const LayoutRoot = memo(
  forwardRef((props, forwardedRef) => {
    const { children, className: rootClassName, ...rootProps } = props;
    const [behavioralTrackingProps, restRootProps] = useBehavioralTrackingProps(rootProps);
    const headerHost = useStableHost();
    const sidebarHost = useStableHost();
    const contentHost = useStableHost();
    const detailsHost = useStableHost();
    const registry = useMemo(() => createSlotRegistry(), []);
    const context = useMemo(
      () => ({
        hosts: {
          header: headerHost,
          sidebar: sidebarHost,
          content: contentHost,
          details: detailsHost
        },
        registry
      }),
      // hosts and registry are stable
      // eslint-disable-next-line react-hooks/exhaustive-deps
      []
    );
    return /* @__PURE__ */ jsx(PageLayoutPortalContext.Provider, { value: context, children: /* @__PURE__ */ jsxs(
      "div",
      {
        ...restRootProps,
        ...behavioralTrackingProps,
        ref: forwardedRef,
        className: clsx(styles.layoutRoot, rootClassName),
        children: [
          /* @__PURE__ */ jsx(HeaderShell, {}),
          /* @__PURE__ */ jsxs("div", { className: styles.contentContainer, children: [
            /* @__PURE__ */ jsx(MainShell, {}),
            children
          ] })
        ]
      }
    ) });
  })
);
LayoutRoot.displayName = "PageLayout";
const PageLayout = Object.assign(LayoutRoot, {
  /** Displays content at the very top of the page. */
  Header: PageLayoutHeader,
  /** Displays content in the collapsible, resizable sidebar panel on the left. */
  Sidebar: PageLayoutSidebar,
  /**
   * Displays the styled main content area (padding, background, border-radius).
   */
  Content: PageLayoutContent,
  /** Displays content in the collapsible, resizable details panel on the right. */
  Details: PageLayoutDetailsWithControlBar
});
export {
  PageLayout,
  usePageLayoutPortalContext
};
