// @dynatrace/strato-components/esm/content/message-container/MessageContainer.js
import { jsx, jsxs } from "react/jsx-runtime";
import { forwardRef } from "react";
import { useIntl } from "react-intl";
import { XmarkIcon } from "@dynatrace/strato-icons";
import { Actions } from "./Actions.js";
import { MessageContainerContext } from "./contexts/MessageContainerContext.js";
import { Description } from "./Description.js";
import { Prefix } from "./Prefix.js";
import { Title } from "./Title.js";
import { Button } from "../../buttons/button/Button.js";
import { useAriaLabelingProps } from "../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useSafeForwardProps } from "../../core/hooks/useSafeForwardProps.js";
import { useSlots } from "../../core/hooks/useSlots.js";
import { devLog } from "../../core/utils/logging.js";
import { Container } from "../../layouts/container/Container.js";
import { Flex } from "../../layouts/flex/Flex.js";
const COMPONENT_NAME = "MessageContainer";
const _MessageContainer = forwardRef(
  (props, forwardedRef) => {
    const {
      children,
      variant = "neutral",
      onDismiss,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-allow": dataDtrumAllow,
      "data-dtrum-mask": dataDtrumMask,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
    const ariaLabelingProps = useAriaLabelingProps(restProps);
    const forwardProps = useSafeForwardProps(restProps, ariaLabelingProps);
    const intl = useIntl();
    const [slots] = useSlots(children, {
      prefix: Prefix,
      title: Title,
      description: Description,
      actions: Actions
    });
    if (!slots.description) {
      devLog(
        "The MessageContainer component requires a Description component."
      );
    }
    const wrapActions = slots.title && slots.description;
    return /* @__PURE__ */ jsx(MessageContainerContext.Provider, { value: { variant }, children: /* @__PURE__ */ jsx(
      Container,
      {
        color: variant,
        padding: 6,
        paddingLeft: 16,
        ref: forwardedRef,
        className: consumerClassName,
        style: consumerStyle,
        "data-testid": dataTestId,
        "data-dtrum-allow": dataDtrumAllow,
        "data-dtrum-mask": dataDtrumMask,
        ...ariaLabelingProps,
        ...behavioralTrackingProps,
        ...forwardProps,
        "data-dt-component": COMPONENT_NAME,
        children: /* @__PURE__ */ jsxs(Flex, { justifyContent: "space-between", alignItems: "flex-start", children: [
          slots.prefix && /* @__PURE__ */ jsx(Flex, { flex: "0 0 20px", paddingY: 6, justifyContent: "center", children: slots.prefix }),
          /* @__PURE__ */ jsxs(
            Flex,
            {
              flexGrow: 1,
              columnGap: 8,
              rowGap: 6,
              flexDirection: wrapActions ? "column" : "row",
              flexWrap: "wrap",
              minWidth: 0,
              children: [
                /* @__PURE__ */ jsxs(
                  Flex,
                  {
                    flexDirection: "column",
                    gap: 2,
                    flexGrow: 1,
                    justifyContent: "center",
                    paddingY: 6,
                    paddingRight: 8,
                    maxWidth: "100%",
                    minWidth: 0,
                    style: { boxSizing: "border-box" },
                    children: [
                      slots.title,
                      slots.description
                    ]
                  }
                ),
                slots.actions && /* @__PURE__ */ jsx(
                  Flex,
                  {
                    ...wrapActions && {
                      style: { paddingRight: 8, paddingBottom: 10 }
                    },
                    children: slots.actions
                  }
                )
              ]
            }
          ),
          onDismiss !== void 0 && /* @__PURE__ */ jsx(Flex, { flexShrink: 0, alignItems: "flex-start", children: /* @__PURE__ */ jsx(
            Button,
            {
              onClick: onDismiss,
              color: variant,
              "aria-label": intl.formatMessage({
                defaultMessage: "Close message",
                id: "2+ypcBK9NuQ5zg1k",
                description: "Aria label used for the close button of the information"
              }),
              children: /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(XmarkIcon, {}) })
            }
          ) })
        ] })
      }
    ) });
  }
);
_MessageContainer.displayName = COMPONENT_NAME;
const MessageContainer = Object.assign(_MessageContainer, {
  Prefix,
  Title,
  Description,
  Actions
});
export {
  MessageContainer
};
