// @dynatrace/strato-components/esm/tables/DataTable/hooks/useTable/createTable.js
import {
  isDev,
  isProduction
} from "../../../../core/utils/is-node-environment.js";
import { DataTableColumnFontStyle } from "../../features/ColumnFontStyle/ColumnFontStyle.js";
import { DataTableColumnOrder } from "../../features/ColumnOrder/ColumnOrder.js";
import { DataTableColumnPinning } from "../../features/ColumnPinning/ColumnPinning.js";
import { DataTableColumnSizing } from "../../features/ColumnSizing/ColumnSizing.js";
import { DataTableColumnVisibility } from "../../features/ColumnVisibility/ColumnVisibility.js";
import { DataTableDownloadData } from "../../features/Download/Download.js";
import { DataTableGanttChartFeature } from "../../features/GanttChart/GanttChart.js";
import { Layout } from "../../features/Layout/Layout.js";
import { LineWrap } from "../../features/LineWrap/LineWrap.js";
import { DataTablePaginationFeature } from "../../features/Pagination/Pagination.js";
import { DataTableRowDetails } from "../../features/RowDetails/RowDetails.js";
import { DataTableRowInteractivity } from "../../features/RowInteractivity/RowInteractivity.js";
import { DataTableRowOrder } from "../../features/RowOrder/RowOrder.js";
import { DataTableRowSelection } from "../../features/RowSelection/RowSelection.js";
import { Sorting as DataTableSorting } from "../../features/Sorting/Sorting.js";
import { DataTableSubRows } from "../../features/SubRows/SubRows.js";
import { Thresholds } from "../../features/Thresholds/Thresholds.js";
import { UserActions } from "../../features/UserActions/UserActions.js";
import { DataTableVirtualScrolling } from "../../features/VirtualScrolling/VirtualScrolling.js";
import { VisualVariantConfig } from "../../features/VisualVariants/VisualVariants.js";
import {
  ColumnFaceting,
  ColumnFiltering,
  ColumnGrouping,
  ColumnOrdering,
  ColumnPinning,
  ColumnSizing,
  ColumnVisibility,
  createColumn,
  functionalUpdate,
  GlobalFaceting,
  GlobalFiltering,
  Headers,
  memo,
  RowExpanding,
  RowPagination,
  RowSelection,
  RowSorting
} from "../../tanstack-table-core-override/src/index.js";
const features = [
  Headers,
  ColumnVisibility,
  ColumnOrdering,
  ColumnPinning,
  ColumnFaceting,
  ColumnFiltering,
  GlobalFaceting,
  //depends on ColumnFaceting
  GlobalFiltering,
  //depends on ColumnFiltering
  RowSorting,
  ColumnGrouping,
  RowExpanding,
  RowPagination,
  RowSelection,
  ColumnSizing,
  DataTableColumnSizing,
  LineWrap,
  Layout,
  VisualVariantConfig,
  UserActions,
  DataTableSorting,
  DataTableColumnVisibility,
  DataTableColumnOrder,
  DataTableRowSelection,
  Thresholds,
  DataTableSubRows,
  DataTableRowDetails,
  DataTableRowInteractivity,
  DataTableVirtualScrolling,
  DataTableDownloadData,
  DataTableColumnFontStyle,
  DataTablePaginationFeature,
  DataTableRowOrder,
  DataTableColumnPinning,
  DataTableGanttChartFeature
];
function createTable(options) {
  if (options.debugAll || options.debugTable) {
    console.info("Creating Table Instance...");
  }
  const table = { _features: features };
  const defaultOptions = table._features.reduce((obj, feature) => {
    return Object.assign(obj, feature.getDefaultOptions?.(table));
  }, {});
  const mergeOptions = (options2) => {
    if (table.options.mergeOptions) {
      return table.options.mergeOptions(defaultOptions, options2);
    }
    return {
      ...defaultOptions,
      ...options2
    };
  };
  const coreInitialState = {};
  let initialState = {
    ...coreInitialState,
    ...options.initialState ?? {}
  };
  table._features.forEach((feature) => {
    initialState = feature.getInitialState?.(initialState) ?? initialState;
  });
  const queued = [];
  let queuedTimeout = false;
  const coreInstance = {
    _features: features,
    options: {
      ...defaultOptions,
      ...options
    },
    initialState,
    _queue: (cb) => {
      queued.push(cb);
      if (!queuedTimeout) {
        queuedTimeout = true;
        Promise.resolve().then(() => {
          while (queued.length) {
            queued.shift()?.();
          }
          queuedTimeout = false;
        }).catch(
          (error) => setTimeout(() => {
            throw error;
          })
        );
      }
    },
    reset: () => {
      table.setState(table.initialState);
    },
    setOptions: (updater) => {
      const newOptions = functionalUpdate(updater, table.options);
      table.options = mergeOptions(newOptions);
    },
    getState: () => {
      return table.options.state;
    },
    setState: (updater) => {
      table.options.onStateChange?.(updater);
    },
    _getRowId: (row, index, parent) => table.options.getRowId?.(row, index, parent) ?? `${parent ? [parent.id, index].join(".") : index}`,
    getCoreRowModel: () => {
      table._getCoreRowModel ??= table.options.getCoreRowModel(table);
      return table._getCoreRowModel();
    },
    // The final calls start at the bottom of the model,
    // row ordering, which then work their way up
    getRowModel: () => {
      return table.getRowDetailsRowModel();
    },
    getRow: (id, searchAll) => {
      const row = (searchAll ? table.getCoreRowModel() : table.getRowModel()).rowsById[id];
      if (!row) {
        if (!isProduction()) {
          throw new Error(`getRow expected an ID, but got ${id}`);
        }
        throw new Error();
      }
      return row;
    },
    _getDefaultColumnDef: memo(
      () => [table.options.defaultColumn],
      (defaultColumn) => {
        defaultColumn = defaultColumn ?? {};
        return {
          header: (props) => {
            const resolvedColumnDef = props.header.column.columnDef;
            if (resolvedColumnDef.accessorKey) {
              return resolvedColumnDef.accessorKey;
            }
            if (resolvedColumnDef.accessorFn) {
              return resolvedColumnDef.id;
            }
            return null;
          },
          // footer: props => props.header.column.id,
          cell: (props) => props.renderValue()?.toString?.() ?? null,
          ...table._features.reduce((obj, feature) => {
            return Object.assign(obj, feature.getDefaultColumnDef?.());
          }, {}),
          ...defaultColumn
        };
      },
      {
        debug: () => table.options.debugAll ?? table.options.debugColumns,
        key: isDev() && "getDefaultColumnDef"
      }
    ),
    _getColumnDefs: () => table.options.columns,
    getAllColumns: memo(
      () => [table._getColumnDefs()],
      (columnDefs) => {
        const recurseColumns = (columnDefs2, parent, depth = 0) => {
          return columnDefs2.map((columnDef) => {
            const column = createColumn(
              // eslint-disable-next-line @typescript-eslint/prefer-ts-expect-error, @typescript-eslint/ban-ts-comment
              // @ts-ignore Strict mode transition - needs to be investigated
              table,
              columnDef,
              depth,
              parent
            );
            const groupingColumnDef = columnDef;
            column.columns = groupingColumnDef.columns ? recurseColumns(groupingColumnDef.columns, column, depth + 1) : [];
            return column;
          });
        };
        return recurseColumns(columnDefs);
      },
      {
        key: isDev() && "getAllColumns",
        debug: () => table.options.debugAll ?? table.options.debugColumns
      }
    ),
    getAllFlatColumns: memo(
      () => [table.getAllColumns()],
      (allColumns) => {
        return allColumns.flatMap((column) => {
          return column.getFlatColumns();
        });
      },
      {
        key: isDev() && "getAllFlatColumns",
        debug: () => table.options.debugAll ?? table.options.debugColumns
      }
    ),
    getAllFlatColumnIdsWithoutBuiltin: memo(
      () => [table.getAllColumns()],
      () => {
        return table.getAllFlatColumns().filter((leafColumn) => !leafColumn.columnDef.meta?.builtin).map((column) => column.id);
      },
      {
        key: isDev() && "getAllFlatColumnIdsWithoutBuiltin",
        debug: () => table.options.debugAll ?? table.options.debugColumns
      }
    ),
    _getAllFlatColumnsById: memo(
      () => [table.getAllFlatColumns()],
      (flatColumns) => {
        return flatColumns.reduce(
          (acc, column) => {
            acc[column.id] = column;
            return acc;
          },
          {}
        );
      },
      {
        key: isDev() && "getAllFlatColumnsById",
        debug: () => table.options.debugAll ?? table.options.debugColumns
      }
    ),
    getAllLeafColumns: memo(
      () => [table.getAllColumns(), table._getOrderColumnsFn()],
      (allColumns, orderColumns) => {
        const leafColumns = allColumns.flatMap(
          (column) => column.getLeafColumns()
        );
        return orderColumns(leafColumns);
      },
      {
        key: isDev() && "getAllLeafColumns",
        debug: () => table.options.debugAll ?? table.options.debugColumns
      }
    ),
    getAllLeafColumnsWithoutBuiltin: memo(
      () => [table.getAllLeafColumns()],
      (allLeafColumns) => {
        return allLeafColumns.filter(
          (column) => !column.columnDef.meta?.builtin
        );
      },
      {
        key: isDev() && "getAllLeafColumnsWithoutBuiltin",
        debug: () => table.options.debugAll ?? table.options.debugColumns
      }
    ),
    getColumn: (columnId) => {
      const column = table._getAllFlatColumnsById()[columnId];
      if (!isProduction() && !column) {
        console.warn(`[Table] Column with id '${columnId}' does not exist.`);
      }
      return column;
    },
    getHeader: memo(
      (headerId) => [headerId, table.getLeafHeaders()],
      (headerId, leafHeaders) => {
        const header = leafHeaders.find((header2) => header2.id === headerId);
        if (!isProduction() && !header) {
          console.warn(`[Table] Header with id '${headerId}' does not exist.`);
        }
        return header;
      },
      {
        key: isDev() && "getHeader",
        debug: () => table.options.debugAll ?? table.options.debugHeaders
      }
    )
  };
  Object.assign(table, coreInstance);
  for (const feature of table._features) {
    feature?.createTable?.(table);
  }
  return table;
}
export {
  createTable
};
