// @dynatrace/strato-components/forms/calendar/calendar-header/CalendarHeader.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CalendarHeader_exports = {};
__export(CalendarHeader_exports, {
  CalendarHeader: () => CalendarHeader
});
module.exports = __toCommonJS(CalendarHeader_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_Button = require("../../../buttons/button/Button.js");
var import_Flex = require("../../../layouts/flex/Flex.js");
var import_useFormFieldContext = require("../../forms-core/hooks/useFormFieldContext.js");
var import_SelectDateLabel = require("../calendar-header-label/SelectDateLabel.js");
var import_SelectMonthLabel = require("../calendar-header-label/SelectMonthLabel.js");
var import_SelectYearLabel = require("../calendar-header-label/SelectYearLabel.js");
var itemStyles = __toESM(require("../components/CalendarItem.sty.js"));
var import_useCalendarContext = require("../hooks/useCalendarContext.js");
var import_calendarConfiguration = require("../types/calendarConfiguration.js");
function getViewSpecificProps(view) {
  switch (view) {
    case "day": {
      const nextView = "month";
      return {
        label: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_SelectDateLabel.SelectDateLabel, {}),
        nextView,
        headerDataDtProperties: `value:${nextView}`,
        payload: { amount: 1, unit: nextView }
      };
    }
    case "month": {
      const nextView = "year";
      return {
        label: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_SelectMonthLabel.SelectMonthLabel, {}),
        nextView,
        headerDataDtProperties: `value:${nextView}`,
        payload: { amount: 1, unit: nextView }
      };
    }
    case "year": {
      const nextView = "day";
      return {
        label: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_SelectYearLabel.SelectYearLabel, {}),
        nextView,
        headerDataDtProperties: `value:${nextView}`,
        payload: { amount: import_calendarConfiguration.NUMBER_OF_YEARS }
      };
    }
  }
}
const CalendarHeader = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const { labelId, view } = props;
  const { dispatch } = (0, import_useCalendarContext.useCalendarContext)();
  const { disabled, ariaDisabled } = (0, import_useFormFieldContext.useFormFieldContext)();
  const intl = (0, import_react_intl.useIntl)();
  const { label, nextView, payload, headerDataDtProperties } = getViewSpecificProps(view);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { justifyContent: "space-between", marginBottom: 16, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Button.Button,
      {
        disabled,
        "aria-disabled": ariaDisabled,
        "aria-label": intl.formatMessage({
          defaultMessage: "Go to previous page.",
          id: "VoqQWp9yEM0QM8eF",
          description: "Aria label to add to the `previous` button"
        }),
        onClick: () => dispatch({
          type: "navigateCurrentView",
          payload: { ...payload, amount: payload.amount * -1 }
        }),
        "data-dt-name": "PreviousPage",
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ChevronLeftSmallIcon, {}) })
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Button.Button,
      {
        className: itemStyles.headerButton,
        disabled,
        "aria-disabled": ariaDisabled,
        "aria-live": "polite",
        id: labelId,
        onClick: () => dispatch({
          type: "currentViewChange",
          payload: nextView
        }),
        ref: forwardedRef,
        "data-dt-name": "ChangeView",
        "data-dt-properties": headerDataDtProperties,
        children: label
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Button.Button,
      {
        disabled,
        "aria-disabled": ariaDisabled,
        "aria-label": intl.formatMessage({
          defaultMessage: "Go to next page.",
          id: "KAAzM9PVtiJno9ks",
          description: "Aria label to add to the `next` button"
        }),
        onClick: () => dispatch({
          type: "navigateCurrentView",
          payload
        }),
        "data-dt-name": "NextPage",
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ChevronRightSmallIcon, {}) })
      }
    )
  ] });
});
CalendarHeader.displayName = "CalendarHeader";
