// @dynatrace/strato-components/charts/core/components/single-value-renderer/constants.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var constants_exports = {};
__export(constants_exports, {
  DEFAULT_MAX_VALUE_FONT_SIZE: () => DEFAULT_MAX_VALUE_FONT_SIZE,
  DEFAULT_SINGLE_VALUE_COLOR: () => DEFAULT_SINGLE_VALUE_COLOR,
  DOWNWARD_ICON_DEFAULT_COLOR: () => DOWNWARD_ICON_DEFAULT_COLOR,
  DOWNWARD_VALUE_DEFAULT_COLOR: () => DOWNWARD_VALUE_DEFAULT_COLOR,
  FONT_HEIGHT_RATIO: () => FONT_HEIGHT_RATIO,
  FONT_WIDTH_RATIO: () => FONT_WIDTH_RATIO,
  GLOBAL_LABEL_FONT_CSS_TOKEN: () => GLOBAL_LABEL_FONT_CSS_TOKEN,
  GLOBAL_TREND_FONT_CSS_TOKEN: () => GLOBAL_TREND_FONT_CSS_TOKEN,
  GLOBAL_VALUE_FONT_CSS_TOKEN: () => GLOBAL_VALUE_FONT_CSS_TOKEN,
  LABEL_FONT_CSS_TOKEN: () => LABEL_FONT_CSS_TOKEN,
  LABEL_FONT_SIZE: () => LABEL_FONT_SIZE,
  MINIMUM_HEIGHT_FOR_TOOLBAR: () => MINIMUM_HEIGHT_FOR_TOOLBAR,
  MINIMUM_HEIGHT_FOR_TOOLBAR_WITH_PADDING: () => MINIMUM_HEIGHT_FOR_TOOLBAR_WITH_PADDING,
  NEUTRAL_ICON_DEFAULT_COLOR: () => NEUTRAL_ICON_DEFAULT_COLOR,
  NEUTRAL_VALUE_DEFAULT_COLOR: () => NEUTRAL_VALUE_DEFAULT_COLOR,
  TREND_FONT_CSS_TOKEN: () => TREND_FONT_CSS_TOKEN,
  TREND_FONT_SIZE: () => TREND_FONT_SIZE,
  TREND_LABEL_DEFAULT_COLOR: () => TREND_LABEL_DEFAULT_COLOR,
  UPWARD_ICON_DEFAULT_COLOR: () => UPWARD_ICON_DEFAULT_COLOR,
  UPWARD_VALUE_DEFAULT_COLOR: () => UPWARD_VALUE_DEFAULT_COLOR,
  VALUE_FONT_CSS_TOKEN: () => VALUE_FONT_CSS_TOKEN,
  VALUE_FONT_SIZE: () => VALUE_FONT_SIZE
});
module.exports = __toCommonJS(constants_exports);
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
const FONT_HEIGHT_RATIO = 0.7;
const FONT_WIDTH_RATIO = 0.95;
const VALUE_FONT_CSS_TOKEN = "--dt-charts-single-value-font-size";
const LABEL_FONT_CSS_TOKEN = "--dt-charts-single-value-label-font-size";
const TREND_FONT_CSS_TOKEN = "--dt-charts-single-value-label-font-size";
const DEFAULT_MAX_VALUE_FONT_SIZE = 240;
const DEFAULT_SINGLE_VALUE_COLOR = import_colors.default.Text.Neutral.Default;
const GLOBAL_VALUE_FONT_CSS_TOKEN = "--dt-charts-single-value-global-value-font-size";
const GLOBAL_LABEL_FONT_CSS_TOKEN = "--dt-charts-single-value-global-label-font-size";
const GLOBAL_TREND_FONT_CSS_TOKEN = "--dt-charts-single-value-global-trend-font-size";
const MAX_VALUE_FONT_SIZE = `var(${GLOBAL_VALUE_FONT_CSS_TOKEN}, ${DEFAULT_MAX_VALUE_FONT_SIZE})`;
const VALUE_FONT_SIZE = `min(var(${VALUE_FONT_CSS_TOKEN}, ${MAX_VALUE_FONT_SIZE}), ${MAX_VALUE_FONT_SIZE})`;
const MAX_LABEL_FONT_SIZE = `min(var(${GLOBAL_LABEL_FONT_CSS_TOKEN}, ${VALUE_FONT_SIZE}), ${VALUE_FONT_SIZE})`;
const LABEL_FONT_SIZE = `min(var(${LABEL_FONT_CSS_TOKEN}, ${MAX_LABEL_FONT_SIZE}), ${MAX_LABEL_FONT_SIZE})`;
const MAX_TREND_FONT_SIZE = `min(var(${GLOBAL_TREND_FONT_CSS_TOKEN}, ${VALUE_FONT_SIZE}), ${VALUE_FONT_SIZE})`;
const TREND_FONT_SIZE = `min(var(${LABEL_FONT_CSS_TOKEN}, ${MAX_TREND_FONT_SIZE}), ${MAX_TREND_FONT_SIZE})`;
const UPWARD_ICON_DEFAULT_COLOR = import_colors.default.Icon.Success.Default;
const NEUTRAL_ICON_DEFAULT_COLOR = import_colors.default.Icon.Primary.Default;
const DOWNWARD_ICON_DEFAULT_COLOR = import_colors.default.Icon.Critical.Default;
const UPWARD_VALUE_DEFAULT_COLOR = import_colors.default.Text.Success.Default;
const NEUTRAL_VALUE_DEFAULT_COLOR = import_colors.default.Text.Primary.Default;
const DOWNWARD_VALUE_DEFAULT_COLOR = import_colors.default.Text.Critical.Default;
const TREND_LABEL_DEFAULT_COLOR = import_colors.default.Text.Neutral.Default;
const MINIMUM_HEIGHT_FOR_TOOLBAR = 40;
const MINIMUM_HEIGHT_FOR_TOOLBAR_WITH_PADDING = 48;
