// @dynatrace/strato-components/charts/core/components/single-value-renderer/SingleValueRenderer.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SingleValueRenderer_exports = {};
__export(SingleValueRenderer_exports, {
  SingleValueRenderer: () => SingleValueRenderer
});
module.exports = __toCommonJS(SingleValueRenderer_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_ContentGrid = require("./components/ContentGrid.js");
var import_Label = require("./components/Label.js");
var import_SingleValueToolbar = require("./components/SingleValueToolbar.js");
var import_Trend = require("./components/trend/Trend.js");
var import_Value = require("./components/Value.js");
var import_single_value_context = require("./contexts/single-value.context.js");
var import_font_size_provider = require("./providers/font-size.provider.js");
var import_SingleValueRenderer_css = require("./SingleValueRenderer.sty.js");
var import_build_value_content_grid_rows = require("./utils/build-value-content-grid-rows.js");
var import_format_label = require("./utils/format-label.js");
var import_format_value = require("./utils/format-value.js");
var import_get_calculated_color = require("./utils/get-calculated-color.js");
var import_get_content_and_background_color = require("./utils/get-content-and-background-color.js");
var import_useContainerSize = require("../../../../core/hooks/useContainerSize.js");
var import_defaultChartValues_css = require("../../../../core/styles/defaultChartValues.sty.js");
var import_ChartStateHandler = require("../chart-states/ChartStateHandler.js");
var import_LoaderOverlay = require("../loader-overlay/LoaderOverlay.js");
var import_SparklineDisplay = require("../sparkline-display/SparklineDisplay.js");
const SingleValueRenderer = (props) => {
  const {
    data,
    formatter,
    unit,
    prefixIcon,
    alignment,
    label,
    color,
    colorRules = [],
    emptyState,
    errorState,
    loading = false,
    sparklineProps,
    sparklineChildren,
    trendProps,
    applyThresholdBackground,
    propsFromGrid,
    toolbar,
    dataItem
  } = props;
  const containerRef = (0, import_react.useRef)(null);
  const { height: containerHeight } = (0, import_useContainerSize.useContainerSize)(containerRef);
  const { color: calculatedColor, ruleApplied } = (0, import_get_calculated_color.getCalculatedColor)(
    data,
    colorRules,
    color,
    dataItem
  );
  const applyBackgroundColor = applyThresholdBackground && ruleApplied;
  const { contentColor, backgroundColor } = (0, import_get_content_and_background_color.getContentAndBackgroundColor)(
    calculatedColor,
    applyBackgroundColor
  );
  const { value, unit: formatterUnit } = (0, import_format_value.formatValue)(data, unit, formatter);
  const formattedLabel = (0, import_format_label.formatLabel)(label);
  const context = {
    value,
    unit: formatterUnit ?? "",
    label: formattedLabel ?? "",
    prefixIcon,
    alignment
  };
  const { innerGrid, outerGrid } = (0, import_build_value_content_grid_rows.buildValueContentGridRows)(
    !(0, import_lodash_es.isUndefined)(label),
    !(0, import_lodash_es.isUndefined)(data),
    !(0, import_lodash_es.isUndefined)(trendProps),
    !(0, import_lodash_es.isUndefined)(sparklineProps)
  );
  const getSparklineColor = () => applyBackgroundColor ? contentColor : sparklineProps?.color ?? contentColor;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      className: import_SingleValueRenderer_css.singleValueContainer,
      style: { backgroundColor },
      ref: containerRef,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_font_size_provider.FontSizeProvider,
        {
          text: context.prefixIcon ? `00${context.value}` : `${context.value}`,
          suffix: context.unit,
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_single_value_context.SingleValueContext.Provider, { value: context, children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_ContentGrid.ContentGrid, { color: contentColor, gridTemplate: outerGrid, children: [
              label !== void 0 ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Label.Label, { color: applyBackgroundColor ? contentColor : void 0 }) : null,
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: (0, import_clsx.default)(import_SingleValueRenderer_css.singleValueInnerContainer), children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
                import_ChartStateHandler.ChartStateHandler,
                {
                  empty: (0, import_lodash_es.isEmpty)(value),
                  height: "100cqh",
                  emptyState,
                  errorState,
                  isLoading: loading,
                  style: {
                    gridTemplate: innerGrid
                  },
                  className: (0, import_clsx.default)(import_defaultChartValues_css.defaultChartSize, import_SingleValueRenderer_css.chartStateHandler),
                  children: [
                    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Value.Value, {}),
                    !(0, import_lodash_es.isUndefined)(trendProps) ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: import_SingleValueRenderer_css.trendArea, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                      import_Trend.Trend,
                      {
                        ...trendProps,
                        contrastColor: applyBackgroundColor ? contentColor : void 0
                      }
                    ) }) : null,
                    !(0, import_lodash_es.isUndefined)(sparklineProps) ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: import_SingleValueRenderer_css.sparklineArea, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                      import_SparklineDisplay.SparklineDisplay,
                      {
                        ...sparklineProps,
                        color: getSparklineColor(),
                        labelsAlignment: "right",
                        labelsFixedWidth: void 0,
                        ticksColor: applyBackgroundColor ? contentColor : void 0,
                        style: {
                          display: "flex",
                          flexDirection: "column",
                          justifyContent: "flex-end"
                        },
                        children: sparklineChildren
                      }
                    ) }) : null,
                    loading && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_LoaderOverlay.LoaderOverlay, {})
                  ]
                }
              ) })
            ] }),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_SingleValueToolbar.SingleValueToolbar,
              {
                ...propsFromGrid?.toolbar ? propsFromGrid?.toolbar : toolbar,
                parentHeight: containerHeight,
                intentsAsyncTemplate: propsFromGrid?.intentsAsyncTemplate
              }
            )
          ] })
        }
      )
    }
  );
};
