// @dynatrace/strato-components/content/empty-state/slot-components/Actions.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Actions_exports = {};
__export(Actions_exports, {
  Actions: () => Actions
});
module.exports = __toCommonJS(Actions_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_Flex = require("../../../layouts/flex/Flex.js");
var import_EmptyState_css = require("../EmptyState.sty.js");
const Actions = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    children,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId
  } = props;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Flex.Flex,
    {
      ref: forwardedRef,
      flexDirection: "row",
      alignItems: "center",
      gap: 8,
      className: (0, import_clsx.default)([import_EmptyState_css.actionsSlotCss, consumerClassName]),
      style: consumerStyle,
      "data-testid": dataTestId,
      children
    }
  );
});
Actions.displayName = "EmptyState.Actions";
