// @dynatrace/strato-components/content/empty-state/slot-components/VisualPreset.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var VisualPreset_exports = {};
__export(VisualPreset_exports, {
  VisualPreset: () => VisualPreset
});
module.exports = __toCommonJS(VisualPreset_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_useCurrentTheme = require("../../../core/hooks/useCurrentTheme.js");
var import_empty_state_utils = require("../empty-state-utils.js");
const VisualPreset = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    context,
    type,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId
  } = props;
  const currentTheme = (0, import_useCurrentTheme.useCurrentTheme)();
  const visualPresetUrl = (0, import_empty_state_utils.useVisualPresetCdnUrl)(context, type, currentTheme);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "img",
    {
      "aria-hidden": true,
      alt: "",
      ref: forwardedRef,
      src: visualPresetUrl,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId
    }
  );
});
VisualPreset.displayName = "EmptyState.VisualPreset";
