// @dynatrace/strato-components/tables/DataTable/components/Toolbar/ColumnSettings/ColumnOrderSettings/ColumnOrderSettingsDragAndDrop.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ColumnOrderSettingsDragAndDrop_exports = {};
__export(ColumnOrderSettingsDragAndDrop_exports, {
  ColumnOrderSettingsDragAndDropContext: () => ColumnOrderSettingsDragAndDropContext
});
module.exports = __toCommonJS(ColumnOrderSettingsDragAndDrop_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_core = require("@dnd-kit/core");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_column_order_settings_dnd_utils = require("./column-order-settings-dnd-utils.js");
var styles = __toESM(require("./ColumnOrderSettingsDragAndDrop.sty.js"));
var import_logging = require("../../../../../../core/utils/logging.js");
const SCROLL_JUMP_THRESHOLD = import_column_order_settings_dnd_utils.DATA_TABLE_COLUMNS_LIST_ITEM_MIN_HEIGHT * 10;
function ColumnOrderSettingsDragAndDropContext({
  children,
  localColumnOrder,
  updateLocalColumnOrder,
  localColumnPinning,
  itemVirtualizer
}) {
  const [dragOverlayContent, setDragOverlayContent] = (0, import_react.useState)(null);
  const [dragOverlaySize, setDragOverlaySize] = (0, import_react.useState)({ width: 0, height: 0 });
  const [draggingColumnInfo, setDraggingColumnInfo] = (0, import_react.useState)(null);
  const modalContentRef = (0, import_react.useRef)(null);
  const scrollHandlerRef = (0, import_react.useRef)(null);
  const sensors = (0, import_core.useSensors)(
    (0, import_core.useSensor)(import_core.MouseSensor, {}),
    (0, import_core.useSensor)(import_core.TouchSensor, {}),
    (0, import_core.useSensor)(import_core.KeyboardSensor, {})
  );
  const columnsListCollisionDetection = (0, import_react.useMemo)(
    () => (0, import_column_order_settings_dnd_utils.createColumnOrderSettingsCollisionDetection)(
      localColumnOrder,
      localColumnPinning,
      draggingColumnInfo
    ),
    [draggingColumnInfo, localColumnOrder, localColumnPinning]
  );
  const restrictToColumnsListAreaModifier = (0, import_react.useCallback)(
    (args) => {
      return (0, import_column_order_settings_dnd_utils.createRestrictToColumnsListAreaModifier)(
        itemVirtualizer.scrollElement
      )(args);
    },
    [itemVirtualizer]
  );
  const resetDragging = () => {
    const scrollElement = itemVirtualizer.scrollElement;
    if (scrollElement && scrollHandlerRef.current) {
      scrollElement.removeEventListener("scroll", scrollHandlerRef.current);
      scrollHandlerRef.current = null;
    }
    if (modalContentRef.current) {
      const allColumns = (0, import_column_order_settings_dnd_utils.getAllColumns)(modalContentRef.current);
      (0, import_column_order_settings_dnd_utils.resetAllColumnsStyles)(allColumns);
    }
    modalContentRef.current = null;
    setDraggingColumnInfo(null);
    setDragOverlayContent(null);
    setDragOverlaySize({ width: 0, height: 0 });
  };
  const resetDraggingAndFocusLastActive = (activeId) => {
    requestAnimationFrame(() => {
      (0, import_column_order_settings_dnd_utils.getDragHandleForColumn)(activeId, modalContentRef.current)?.focus();
      resetDragging();
    });
  };
  const handleOnDragStart = ({ active }) => {
    const activeLocalColumn = active.data.current?.column;
    if (!activeLocalColumn) {
      return;
    }
    const scrollElement = itemVirtualizer.scrollElement;
    modalContentRef.current = scrollElement?.firstElementChild || scrollElement;
    const localColumn = (0, import_column_order_settings_dnd_utils.getColumnById)(
      modalContentRef.current,
      activeLocalColumn.id
    );
    if (!localColumn) {
      (0, import_logging.devLog)(
        "[DataTable] Could not find column element in DOM",
        activeLocalColumn.id
      );
      return;
    }
    const columnInfo = {
      id: active.id,
      index: active.data.current?.columnIndex,
      column: activeLocalColumn
    };
    setDraggingColumnInfo(columnInfo);
    let lastScrollOffset = itemVirtualizer.scrollOffset ?? scrollElement?.scrollTop ?? 0;
    const handleScroll = () => {
      const innerContainer = modalContentRef.current;
      if (!innerContainer || !columnInfo.column) {
        return;
      }
      (0, import_column_order_settings_dnd_utils.styleNonTargetColumns)(innerContainer, columnInfo.column.parent?.id);
      (0, import_column_order_settings_dnd_utils.hideChildColumnsOnScroll)(innerContainer);
      const currentOffset = scrollElement.scrollTop;
      const scrollDelta = Math.abs(currentOffset - lastScrollOffset);
      if (scrollDelta >= SCROLL_JUMP_THRESHOLD) {
        scrollElement.scrollTop = lastScrollOffset;
        return;
      }
      lastScrollOffset = currentOffset;
    };
    scrollHandlerRef.current = handleScroll;
    scrollElement?.addEventListener("scroll", handleScroll);
    (0, import_column_order_settings_dnd_utils.styleNonTargetColumns)(
      modalContentRef.current,
      activeLocalColumn.parent?.id
    );
    const dragOverlaySize2 = {
      width: localColumn.clientWidth,
      height: localColumn.clientHeight
    };
    const clonedElement = localColumn.cloneNode(true);
    if (activeLocalColumn.columns.length === 0) {
      setDragOverlayContent(clonedElement);
    } else {
      const wrapper = document.createElement("div");
      wrapper.appendChild(clonedElement);
      const childColumnIds = (0, import_column_order_settings_dnd_utils.getLeafColumnsForColumn)(
        activeLocalColumn,
        localColumnOrder
      );
      for (const childColumnId of childColumnIds) {
        const childColumn = (0, import_column_order_settings_dnd_utils.getColumnById)(
          modalContentRef.current,
          childColumnId
        );
        if (childColumn) {
          const clonedChild = childColumn.cloneNode(true);
          wrapper.appendChild(clonedChild);
          dragOverlaySize2.height += childColumn.clientHeight + import_column_order_settings_dnd_utils.DATA_TABLE_COLUMNS_LIST_ITEM_BOTTOM_BORDER;
        }
      }
      (0, import_column_order_settings_dnd_utils.hideChildColumns)(modalContentRef.current, childColumnIds);
      setDragOverlayContent(wrapper);
    }
    setDragOverlaySize(dragOverlaySize2);
  };
  const handleOnDragOver = (event) => {
    (0, import_column_order_settings_dnd_utils.transformColumnsOnDragOver)(
      modalContentRef.current,
      event,
      localColumnOrder,
      dragOverlaySize.height,
      draggingColumnInfo?.column
    );
  };
  const handleOnDragEnd = ({ over }) => {
    if (!draggingColumnInfo) {
      (0, import_logging.devLog)("[Table] On drag end: dragging column info is undefined");
      resetDragging();
      return;
    }
    const activeColumn = draggingColumnInfo.column;
    const activeId = activeColumn.id;
    if (over === void 0 || over === null) {
      (0, import_logging.devLog)("[Table] On drag end: over info is undefined or null", over);
      resetDraggingAndFocusLastActive(activeId);
      return;
    }
    const overId = over.data.current?.columnId;
    if (overId === activeId) {
      resetDraggingAndFocusLastActive(activeId);
      return;
    }
    const { fromIndex, toIndex } = (0, import_column_order_settings_dnd_utils.getDragIndexes)(
      localColumnOrder,
      activeId,
      overId
    );
    const columnsToMove = [activeId];
    if (activeColumn.columns.length > 0) {
      const childColumnIds = (0, import_column_order_settings_dnd_utils.getLeafColumnsForColumn)(
        activeColumn,
        localColumnOrder
      );
      columnsToMove.push(...childColumnIds);
    }
    const directionOffset = fromIndex < toIndex ? 1 : -1;
    const adjustedTargetIndex = fromIndex < toIndex ? Math.max(0, toIndex - columnsToMove.length + directionOffset) : toIndex;
    const newColumnOrder = [...localColumnOrder];
    newColumnOrder.splice(fromIndex, columnsToMove.length);
    newColumnOrder.splice(adjustedTargetIndex, 0, ...columnsToMove);
    updateLocalColumnOrder(newColumnOrder);
    resetDraggingAndFocusLastActive(activeId);
  };
  const onDragCancel = () => {
    const activeId = draggingColumnInfo?.id;
    if (activeId) {
      resetDraggingAndFocusLastActive(activeId);
    }
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_core.DndContext,
    {
      accessibility: { restoreFocus: false },
      sensors,
      collisionDetection: columnsListCollisionDetection,
      modifiers: [restrictToColumnsListAreaModifier],
      onDragStart: handleOnDragStart,
      onDragOver: handleOnDragOver,
      onDragEnd: handleOnDragEnd,
      onDragCancel,
      children: [
        children,
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_core.DragOverlay,
          {
            dropAnimation: null,
            className: (0, import_clsx.default)(styles.columnDndOverlay),
            style: {
              height: dragOverlaySize.height,
              width: dragOverlaySize.width
            },
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              "div",
              {
                ref: (node) => dragOverlayContent && node?.appendChild(dragOverlayContent)
              }
            )
          }
        )
      ]
    }
  );
}
