// @dynatrace/strato-components/forms/date-time-picker/DateTimePickerRoot.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DateTimePickerRoot_exports = {};
__export(DateTimePickerRoot_exports, {
  DateTimePickerRoot: () => DateTimePickerRoot
});
module.exports = __toCommonJS(DateTimePickerRoot_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_DateSelector = require("./DateSelector.js");
var styles = __toESM(require("./DateTimePicker.sty.js"));
var import_Button = require("../../buttons/button/Button.js");
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useFormControlState = require("../../core/hooks/useFormControlState.js");
var import_useFocusRing = require("../../core/styles/useFocusRing.js");
var import_useTimeframeStore = require("../../core/time/timeframeStore/useTimeframeStore.js");
var import_Flex = require("../../layouts/flex/Flex.js");
var import_InputGroup_css = require("../../layouts/input-group/InputGroup.sty.js");
var import_InputGroup = require("../../layouts/input-group/InputGroup.js");
var import_field_css = require("../../styles/field.sty.js");
var import_DateTimeInput = require("../date-time-range-picker/DateTimeInput.js");
var import_DateTimeRangePicker_css = require("../date-time-range-picker/DateTimeRangePicker.sty.js");
var import_FormFieldMessageBaseItem = require("../forms-core/components/form-field-messages/FormFieldMessageBaseItem.js");
var import_useFormFieldMessagesContext = require("../forms-core/hooks/useFormFieldMessagesContext.js");
var import_useMinMaxLogger = require("../forms-core/hooks/useMinMaxLogger.js");
var import_useSyncedFormControlProps = require("../forms-core/hooks/useSyncedFormControlProps.js");
var import_formfield_messages_state_reducer = require("../forms-core/state/formfield-messages-state-reducer.js");
var import_useInputHint = require("../hooks/useInputHint.js");
var import_focusDateTimeInput = require("./utils/focusDateTimeInput.js");
var import_expression_trigger_helpers = require("../date-time-range-picker/expression-trigger-helpers.js");
var import_helpers = require("../date-time-range-picker/helpers.js");
var import_selection_copy_helpers = require("../date-time-range-picker/selection-copy-helpers.js");
var import_useSelectionInteractionHandlers = require("../date-time-range-picker/useSelectionInteractionHandlers.js");
const COMPONENT_NAME = "DateTimePicker";
const DateTimePickerRoot = (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    id: propId,
    disabled: propDisabled,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    readOnly = false,
    required: propRequired,
    "aria-disabled": ariaDisabledProp,
    placeholder,
    name,
    type = "datetime",
    controlState,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const min = (0, import_useTimeframeStore.useDateTimePickerStore)((state) => state.min);
  const max = (0, import_useTimeframeStore.useDateTimePickerStore)((state) => state.max);
  const value = (0, import_useTimeframeStore.useDateTimePickerStore)((state) => state.dateTimePickerValue);
  const setValue = (0, import_useTimeframeStore.useDateTimePickerStore)(
    (state) => state.setDateTimePickerValue
  );
  const dateInputRef = (0, import_useTimeframeStore.useDateTimePickerStore)(
    (state) => state.dateTimeInputRef
  );
  const rootRef = (0, import_react.useRef)(null);
  const dateSelectorRef = (0, import_react.useRef)(null);
  const [hasInputContent, setHasInputContent] = (0, import_react.useState)(false);
  (0, import_useMinMaxLogger.useMinMaxLogger)(min, max, "DateTimePicker");
  const touchedContainerRef = (0, import_react.useRef)(null);
  const { state: formFieldState, dispatch: formFieldDispatch } = (0, import_useFormFieldMessagesContext.useFormFieldMessagesContext)();
  const { showError: showFormFieldError } = formFieldState;
  const [hasFormControlError, hint] = (0, import_useFormControlState.useFormControlState)(controlState);
  const { hintProps } = (0, import_useInputHint.useInputHint)({
    hint,
    hasError: hasFormControlError
  });
  const showError = hasFormControlError || showFormFieldError;
  const { disabled, required, id, ariaDisabled } = (0, import_useSyncedFormControlProps.useSyncedFormControlProps)({
    propDisabled,
    propRequired,
    propId,
    propAriaDisabled: ariaDisabledProp
  });
  const disabledOrAriaDisabled = disabled || ariaDisabled || false;
  const ariaLabelingProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(remainingProps);
  const showDateSelector = type !== "time";
  const { focusClassName } = (0, import_useFocusRing.useFocusRing)({
    focusWithin: true,
    ignoreModality: true,
    color: showError ? "critical" : "neutral",
    disabled
  });
  const intl = (0, import_react_intl.useIntl)();
  const calendarTriggerLabel = intl.formatMessage({
    defaultMessage: "Datepicker calendar trigger",
    id: "Uuc6xAAPyctyHamM",
    description: "Text shown for the datetimepicker calendar trigger."
  });
  const clearButtonLabel = intl.formatMessage({
    defaultMessage: "Clear date and time",
    id: "5i26Rr1LLz0OIGNJ",
    description: "Aria label for the DateTimePicker clear button."
  });
  (0, import_react.useImperativeHandle)(
    forwardedRef,
    () => ({
      element: rootRef.current,
      inputRef: dateInputRef.current?.inputRef,
      triggerRef: dateInputRef.current?.inputRef,
      open: () => dateSelectorRef.current?.open(),
      close: () => dateSelectorRef.current?.close(),
      focus() {
        dateInputRef.current?.inputRef?.focus();
      },
      validate() {
        dateInputRef.current?.validate();
        formFieldDispatch?.({
          type: import_formfield_messages_state_reducer.FormFieldMessagesAction.SETISTOUCHED,
          payload: { isTouched: true }
        });
      }
    }),
    [formFieldDispatch, dateInputRef]
  );
  const onBlurTouched = (0, import_react.useCallback)(
    (e) => {
      if (!touchedContainerRef.current?.contains(e.relatedTarget)) {
        formFieldDispatch?.({
          type: import_formfield_messages_state_reducer.FormFieldMessagesAction.SETISTOUCHED,
          payload: { isTouched: true }
        });
      }
    },
    [formFieldDispatch]
  );
  (0, import_react.useEffect)(() => {
    if (value === null && dateInputRef.current?.expressionInputRef && dateInputRef.current.expressionInputRef.textContent !== "" && dateInputRef.current.element?.getAttribute("data-valuetype") !== "expression") {
      dateInputRef.current.expressionInputRef.textContent = "";
    }
  }, [value, dateInputRef]);
  const { onMouseDown, onKeyDown: handleSelectionKeyDown } = (0, import_useSelectionInteractionHandlers.useSelectionInteractionHandlers)(
    touchedContainerRef,
    (0, import_react.useCallback)(() => {
      (0, import_focusDateTimeInput.focusDateTimeInput)(null, dateInputRef);
    }, [dateInputRef]),
    dateInputRef
  );
  const onKeyDown = (0, import_react.useCallback)(
    (event) => {
      if (disabled || readOnly) {
        handleSelectionKeyDown(event);
        return;
      }
      const isIso = dateInputRef.current?.element?.getAttribute("data-valuetype") === "iso8601";
      if ((0, import_expression_trigger_helpers.isExpressionTriggerKey)(event) && isIso) {
        dateInputRef.current?.setExpressionValue(event.key);
        requestAnimationFrame(() => {
          const expressionInputRef = dateInputRef.current?.expressionInputRef;
          if (expressionInputRef) {
            expressionInputRef.focus();
            const selection = window.getSelection();
            if (selection) {
              const range = document.createRange();
              range.selectNodeContents(expressionInputRef);
              range.collapse(false);
              selection.removeAllRanges();
              selection.addRange(range);
            }
          }
        });
        event.preventDefault();
        event.stopPropagation();
        return;
      }
      if ((0, import_helpers.isContentModifyingKey)(event)) {
        dateInputRef.current?.clearPasteUndo();
      }
      handleSelectionKeyDown(event);
    },
    [disabled, readOnly, handleSelectionKeyDown, dateInputRef]
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_Flex.Flex,
    {
      gap: 4,
      flexDirection: "column",
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ref: rootRef,
      className: consumerClassName,
      style: consumerStyle,
      ...behavioralTrackingProps,
      "data-dt-component": COMPONENT_NAME,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
          import_InputGroup.InputGroup,
          {
            className: (0, import_clsx.default)(
              styles.dateTimePicker({
                type,
                disabled: disabledOrAriaDisabled,
                showError,
                readOnly
              })
            ),
            style: { gridAutoColumns: "1fr auto" },
            children: [
              /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
                "div",
                {
                  ref: touchedContainerRef,
                  className: (0, import_clsx.default)(
                    (0, import_field_css.fieldTokens)({
                      color: showError ? "critical" : "neutral",
                      disabled: disabledOrAriaDisabled,
                      variant: "emphasized"
                    }),
                    (0, import_DateTimeRangePicker_css.dateTimeRange)({
                      showError,
                      disabled: disabledOrAriaDisabled,
                      readOnly
                    }),
                    focusClassName,
                    import_InputGroup_css.inputGroupClassName
                  ),
                  tabIndex: disabled ? void 0 : 0,
                  onBlur: onBlurTouched,
                  onInput: () => {
                    setHasInputContent(
                      !!dateInputRef.current?.expressionInputRef?.textContent
                    );
                  },
                  onMouseDown,
                  onKeyDown,
                  onFocus: (e) => {
                    if (!rootRef.current?.contains(e.relatedTarget)) {
                      (0, import_focusDateTimeInput.focusDateTimeInput)(e, dateInputRef);
                    }
                  },
                  onCopyCapture: import_selection_copy_helpers.onTimeframeCopyCapture,
                  children: [
                    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                      import_DateTimeInput.DateTimeInput,
                      {
                        ref: dateInputRef,
                        id,
                        type,
                        disabled,
                        readOnly,
                        placeholder,
                        name,
                        value,
                        onChange: (newValue) => {
                          if (typeof newValue !== "string") {
                            setValue(newValue);
                          }
                        },
                        ...ariaLabelingProps,
                        "aria-disabled": ariaDisabled,
                        required,
                        validationValueMode: "single",
                        controlId: id,
                        controlState,
                        onPasteComplete: () => {
                          requestAnimationFrame(() => {
                            dateInputRef.current?.focusLast();
                          });
                        }
                      }
                    ),
                    !readOnly && !disabledOrAriaDisabled && (value !== null && value.value !== "" || hasInputContent) && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                      import_Button.Button,
                      {
                        onFocus: () => {
                          formFieldDispatch?.({
                            type: import_formfield_messages_state_reducer.FormFieldMessagesAction.SETISTOUCHED,
                            payload: { isTouched: true }
                          });
                        },
                        onClick: () => {
                          setValue(null);
                          if (dateInputRef.current?.expressionInputRef) {
                            dateInputRef.current.expressionInputRef.textContent = "";
                          }
                          setHasInputContent(false);
                          requestAnimationFrame(() => {
                            dateInputRef.current?.expressionInputRef?.focus();
                          });
                        },
                        onKeyDown: (e) => {
                          if (e.key === "Enter") {
                            e.stopPropagation();
                          }
                        },
                        className: styles.clearIconButton,
                        color: showError ? "critical" : "neutral",
                        size: "condensed",
                        "aria-label": clearButtonLabel,
                        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.XmarkIcon, {})
                      }
                    )
                  ]
                }
              ),
              showDateSelector && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_DateSelector.DateSelector,
                {
                  ref: dateSelectorRef,
                  "aria-label": calendarTriggerLabel,
                  "aria-disabled": ariaDisabled,
                  readOnly,
                  disabled: !!disabled,
                  error: showError,
                  className: (0, import_clsx.default)(styles.dateSelector, import_InputGroup_css.inputGroupClassName),
                  "data-dtrum-mask": dataDtrumMask,
                  "data-dtrum-allow": dataDtrumAllow
                }
              )
            ]
          }
        ),
        hint && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_FormFieldMessageBaseItem.InternalBaseItem,
          {
            variant: hasFormControlError ? "error" : "hint",
            disabled: disabled || ariaDisabled,
            ...hintProps,
            children: hint
          }
        )
      ]
    }
  );
});
DateTimePickerRoot.displayName = COMPONENT_NAME;
