// @dynatrace/strato-components/forms/date-time-picker/DateTimePicker.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DateTimePicker_exports = {};
__export(DateTimePicker_exports, {
  DateTimePicker: () => DateTimePicker
});
module.exports = __toCommonJS(DateTimePicker_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_DateTimePickerRoot = require("./DateTimePickerRoot.js");
var import_useCallbackRef = require("../../core/hooks/useCallbackRef.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_dateTimePicker = require("../../core/time/timeframeStore/actions/dateTimePicker.js");
var import_createTimeframeStore = require("../../core/time/timeframeStore/createTimeframeStore.js");
var import_timeframeStoreProvider = require("../../core/time/timeframeStore/timeframeStoreProvider.js");
var import_useDateTimeMessages = require("../../core/time/useDateTimeMessages.js");
var import_useMinMax = require("../../core/time/useMinMax.js");
var import_validation = require("../../core/time/validation.js");
var import_helpers = require("../../forms/forms-core/time/helpers.js");
var import_timeParser = require("../../forms/forms-core/time/timeParser.js");
var import_helpers2 = require("../calendar/utils/helpers.js");
var import_useFormFieldMessagesContext = require("../forms-core/hooks/useFormFieldMessagesContext.js");
var import_useSyncedFormControlProps = require("../forms-core/hooks/useSyncedFormControlProps.js");
var import_date_time = require("../forms-core/time/date-time.js");
function useSyncPropsAndState(store, memoedValue, isControlled, precision, formatString, required, formFieldMessagesDispatch, dateTimePickerRef) {
  (0, import_react.useEffect)(() => {
    isControlled && store.setState((state) => {
      if (memoedValue?.absoluteDate === "") {
        return state;
      }
      const error = (0, import_validation.validate)(memoedValue ?? null, state.min, state.max, {
        formMessageDispatch: formFieldMessagesDispatch,
        formerError: state.calendarError,
        precision: state.precision,
        showError: false,
        checkOrder: true,
        ref: dateTimePickerRef,
        expectedFormat: state.formatString,
        valueMode: "single"
      });
      const value = (0, import_timeParser.removePrecisionFromTimeValue)(
        memoedValue ?? null,
        state.precision
      );
      const timeOffset = (0, import_helpers2.getTimeOffsetFromInitialValue)(
        value,
        "single",
        state.min,
        state.max,
        state.precision
      );
      return {
        ...state,
        dateTimePickerValue: value,
        calendarValue: (0, import_helpers.toCalendarValue)(value, state.precision),
        calendarError: error,
        dateTimePickerError: error,
        timeOffset
      };
    });
  }, [
    formFieldMessagesDispatch,
    isControlled,
    memoedValue,
    store,
    dateTimePickerRef
  ]);
  (0, import_react.useEffect)(() => {
    store.setState((state) => {
      const value = (0, import_timeParser.removePrecisionFromTimeValue)(
        state.dateTimePickerValue,
        precision
      );
      const error = (0, import_validation.validate)(value, state.min, state.max, {
        formMessageDispatch: formFieldMessagesDispatch,
        formerError: state.dateTimePickerError,
        precision,
        showError: false,
        checkOrder: true,
        ref: dateTimePickerRef,
        expectedFormat: formatString,
        valueMode: "single",
        required
      });
      return {
        ...state,
        formatString,
        precision,
        dateTimePickerValue: value,
        calendarValue: (0, import_helpers.toCalendarValue)(value, precision),
        required,
        calendarError: error,
        dateTimePickerError: error
      };
    });
  }, [
    formatString,
    precision,
    store,
    required,
    formFieldMessagesDispatch,
    dateTimePickerRef
  ]);
}
const DateTimePicker = (0, import_react.forwardRef)(
  ({
    value: propValue,
    defaultValue: propDefaultValue,
    onChange,
    min: propMin = import_date_time.GRAIL_MIN_DATE,
    max: propMax = import_date_time.GRAIL_MAX_DATE,
    precision: propPrecision = "minutes",
    disabled: propDisabled,
    required: propRequired,
    readOnly: propReadOnly,
    id: propId,
    "aria-disabled": propAriaDisabled,
    type,
    ...remainingProps
  }, forwardedRef) => {
    const [isControlled] = (0, import_react.useState)(propValue !== void 0);
    const onChangeProp = (0, import_useCallbackRef.useCallbackRef)(onChange);
    const precision = type === "date" ? "day" : propPrecision;
    const memoedValue = (0, import_react.useMemo)(
      () => (0, import_timeParser.getTimeValue)(propValue, precision) ?? null,
      [precision, propValue]
    );
    const memoedDefaultValue = (0, import_react.useMemo)(
      () => (0, import_timeParser.getTimeValue)(propDefaultValue, precision),
      /** We're ignoring updates to defaultValue as it only considers the initial state. */
      // eslint-disable-next-line react-hooks/exhaustive-deps
      []
    );
    const { min, max } = (0, import_useMinMax.useMinMax)(propMin, propMax, precision);
    const calendarRef = (0, import_react.useRef)(null);
    const dateTimeInputRef = (0, import_react.useRef)(null);
    const dateTimePickerRef = (0, import_react.useRef)(null);
    const { dispatch: formFieldMessagesDispatch } = (0, import_useFormFieldMessagesContext.useFormFieldMessagesContext)();
    const { disabled, required, id, ariaDisabled, readOnly } = (0, import_useSyncedFormControlProps.useSyncedFormControlProps)({
      propDisabled,
      propRequired,
      propId,
      propAriaDisabled,
      propReadOnly
    });
    const { formatString } = (0, import_useDateTimeMessages.useDateTimeMessages)({ precision });
    const [initialState] = (0, import_react.useState)(() => {
      const _value = isControlled ? memoedValue : memoedDefaultValue;
      const value = _value ?? null;
      const timeOffset = (0, import_helpers2.getTimeOffsetFromInitialValue)(
        value,
        "single",
        min,
        max,
        precision
      );
      const error = (0, import_validation.validate)(value, min, max, {
        formMessageDispatch: formFieldMessagesDispatch,
        precision,
        showError: false,
        checkOrder: true,
        ref: dateTimePickerRef,
        expectedFormat: formatString,
        valueMode: "single"
      });
      const calendarValue = (0, import_helpers.toCalendarValue)(value, precision);
      return {
        calendarValue,
        dateTimePickerValue: value,
        min,
        max,
        precision,
        calendarRef,
        dateTimeInputRef,
        dateTimePickerRef,
        calendarError: error,
        dateTimePickerError: error,
        timeOffset
      };
    });
    const [store] = (0, import_react.useState)(
      () => (0, import_createTimeframeStore.createTimeframeStore)(
        initialState,
        (0, import_dateTimePicker.getDateTimePickerActions)(
          formFieldMessagesDispatch,
          onChangeProp,
          isControlled
        )
      )
    );
    useSyncPropsAndState(
      store,
      memoedValue,
      isControlled,
      precision,
      formatString,
      required,
      formFieldMessagesDispatch,
      dateTimePickerRef
    );
    (0, import_useMinMax.useMinMaxStoreSync)(
      store,
      min,
      max,
      (0, import_react.useCallback)(
        (state, min2, max2) => {
          const error = (0, import_validation.validate)(state.dateTimePickerValue, min2, max2, {
            formMessageDispatch: formFieldMessagesDispatch,
            formerError: state.calendarError,
            precision: state.precision,
            showError: false,
            checkOrder: true,
            ref: dateTimePickerRef,
            expectedFormat: state.formatString,
            valueMode: "single"
          });
          return { calendarError: error, dateTimePickerError: error };
        },
        [formFieldMessagesDispatch]
      )
    );
    const mergedRef = (0, import_useMergeRefs.useMergeRefs)([dateTimePickerRef, forwardedRef]);
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_timeframeStoreProvider.DateTimePickerStoreProvider, { store, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_DateTimePickerRoot.DateTimePickerRoot,
      {
        ...remainingProps,
        id,
        ref: mergedRef,
        "aria-disabled": ariaDisabled,
        disabled,
        required,
        readOnly,
        type
      }
    ) });
  }
);
