// @dynatrace/strato-components/forms/date-time-picker/DateSelector.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DateSelector_exports = {};
__export(DateSelector_exports, {
  DateSelector: () => DateSelector
});
module.exports = __toCommonJS(DateSelector_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_Button = require("../../buttons/button/Button.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_is_visual_refresh_enabled = require("../../core/utils/is-visual-refresh-enabled.js");
var import_useOverlayWithTrigger = require("../../overlays/hooks/useOverlayWithTrigger.js");
var import_Overlay = require("../../overlays/overlay/Overlay.js");
var import_safe_sprinkles = require("../../styles/safe-sprinkles.js");
var import_CalendarRoot = require("../calendar/CalendarRoot.js");
var import_FormFieldMessageContext = require("../forms-core/contexts/FormFieldMessageContext.js");
var import_useFormFieldMessagesContext = require("../forms-core/hooks/useFormFieldMessagesContext.js");
const DateSelector = (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const {
      disabled,
      error,
      readOnly,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "aria-disabled": ariaDisabledProp,
      ...remainingProps
    } = props;
    const [open, setOpen] = (0, import_react.useState)(false);
    const {
      targetRef,
      triggerRef,
      overlayRef,
      overlayState,
      overlayProps,
      overlayTriggerProps
    } = (0, import_useOverlayWithTrigger.useOverlayWithTrigger)({
      placement: "bottom-end",
      offset: 8,
      trapFocus: false,
      shouldCloseOnBlur: false,
      open,
      onOpenChange: setOpen,
      // the overlay should not open when disabled or readOnly
      disabled: disabled || readOnly
    });
    const rootRef = (0, import_useMergeRefs.useMergeRefs)([triggerRef, targetRef]);
    const intl = (0, import_react_intl.useIntl)();
    const calendarInputLabel = intl.formatMessage({
      defaultMessage: "Calendar for picking a date",
      id: "qpf+aqg3hXoviypu",
      description: "Aria label for the DateTimePicker calendar input used for form validation."
    });
    (0, import_react.useImperativeHandle)(
      forwardedRef,
      () => ({
        open: () => setOpen(true),
        close: () => setOpen(false)
      }),
      []
    );
    const { state: formFieldState } = (0, import_useFormFieldMessagesContext.useFormFieldMessagesContext)();
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_Button.Button,
        {
          ref: rootRef,
          "data-testid": dataTestId,
          className: consumerClassName,
          "aria-disabled": ariaDisabledProp,
          style: consumerStyle,
          ...remainingProps,
          ...overlayTriggerProps,
          size: "condensed",
          color: error ? "critical" : "neutral",
          variant: (0, import_is_visual_refresh_enabled.isVisualRefreshEnabled)() ? "emphasized" : "default",
          "data-dt-name": "ShowCalendar",
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.CalendarIcon, {}) })
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_Overlay.Overlay,
        {
          ref: overlayRef,
          isOpen: overlayState.isOpen,
          className: (0, import_safe_sprinkles.globalSprinkles)({
            paddingLeft: 16,
            paddingTop: 16,
            paddingRight: 16,
            paddingBottom: 16
          }),
          overlayProps,
          "data-dtrum-mask": dataDtrumMask,
          "data-dtrum-allow": dataDtrumAllow,
          "aria-label": intl.formatMessage({
            defaultMessage: "Calendar overlay for picking a date",
            id: "A9Pdb5T48IHfEauq",
            description: "Aria label for the date time picker overlay."
          }),
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_FormFieldMessageContext.FormFieldMessageContext.Provider,
            {
              value: {
                state: {
                  customMessages: [],
                  formControlMessages: [],
                  showError: formFieldState.showError
                }
              },
              children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_CalendarRoot.CalendarRoot,
                {
                  "aria-label": calendarInputLabel,
                  "aria-disabled": ariaDisabledProp
                }
              )
            }
          )
        }
      )
    ] });
  }
);
