// @dynatrace/strato-components/esm/charts/pie/components/PieChartInternal.js
import { jsx, jsxs } from "react/jsx-runtime";
import {
  forwardRef,
  useRef
} from "react";
import { useIntl } from "react-intl";
import { PieChartLegend } from "./PieChartLegend/PieChartLegend.js";
import { PieChartOverlays } from "./PieChartOverlays/PieChartOverlays.js";
import { PieChartPlot } from "./PieChartPlot/PieChartPlot.js";
import { useBehavioralTrackingProps } from "../../../core/hooks/useBehavioralTrackingProps.js";
import { useMergeRefs } from "../../../core/hooks/useMergeRefs.js";
import { ChartLayout } from "../../core/components/chart-layout/ChartLayout.js";
import { useAutoLegendRefresh } from "../../core/components/chart-layout/hooks/useAutoLegendRefresh.js";
import { OverlayProvider } from "../../core/components/overlay-tooltip/contexts/OverlayContext.js";
import { TooltipProvider } from "../../core/components/overlay-tooltip/contexts/TooltipContext.js";
import { ORIENTATION } from "../../core/constants.js";
import { sanitizeLegendRatio } from "../../core/utils/sanitize-legend-ratio.js";
import { useChartLegend } from "../context/ChartLegendContext.js";
import { usePieChartRawData } from "../context/PieChartRawDataContext.js";
import { PIE_CHART_MESSAGES } from "../messages.js";
import { PieChartSlicesDataProvider } from "../providers/PieChartSlicesDataProvider.js";
import {
  DEFAULT_CONTAINER_PADDING,
  DEFAULT_SIZE,
  PIE_CATEGORICAL_LEGEND_SIZES
} from "../types/default.js";
const PieChartInternal = forwardRef((props, forwardedRef) => {
  const {
    height,
    width,
    isEmpty,
    emptyState,
    errorState,
    loading = false,
    className,
    style,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "data-testid": dataTestId,
    "aria-label": ariaLabel,
    "data-dt-component": dataDtComponent,
    role
  } = props;
  const [behavioralTrackingProps] = useBehavioralTrackingProps(props);
  const {
    hidden: legendHidden,
    position: legendPosition,
    // eslint-disable-next-line @typescript-eslint/no-deprecated
    ratio,
    initialRatio,
    onRatioChange
  } = useChartLegend();
  const legendRatio = initialRatio ?? ratio;
  const plotAreaRef = useRef(null);
  const mergedRef = useMergeRefs([forwardedRef, plotAreaRef]);
  const intl = useIntl();
  const chartData = usePieChartRawData();
  const isInvalidDataset = chartData.slices?.reduce((acc, slice) => acc + slice.value, 0) <= 0;
  const invalidState = !isEmpty && isInvalidDataset ? emptyState ?? intl.formatMessage(PIE_CHART_MESSAGES.invalidDataset) : emptyState;
  const overrideRatio = (orientation, width2, height2, ratio2, legendContainerDimensions) => {
    const isHorizontal = orientation === ORIENTATION.Horizontal;
    if (!isHorizontal) {
      return legendContainerDimensions;
    }
    const sanitizedRatio = sanitizeLegendRatio(ratio2);
    const legendHeight = isHorizontal ? height2 : sanitizedRatio * height2 * 0.01;
    let legendWidth = width2;
    if (isHorizontal) {
      legendWidth = ratio2 === "auto" ? width2 - height2 : width2 * sanitizedRatio * 0.01;
    }
    return {
      width: legendWidth,
      height: legendHeight
    };
  };
  const chartLayoutRef = useAutoLegendRefresh(chartData);
  return /* @__PURE__ */ jsx(OverlayProvider, { children: /* @__PURE__ */ jsx(TooltipProvider, { children: /* @__PURE__ */ jsx(
    ChartLayout,
    {
      ...behavioralTrackingProps,
      ref: chartLayoutRef,
      chartWidth: width ?? "100%",
      chartHeight: height ?? DEFAULT_SIZE,
      isEmpty: isEmpty || isInvalidDataset,
      emptyState: isEmpty ? emptyState : emptyState ?? invalidState,
      errorState,
      showLoader: loading,
      className,
      style: { padding: `0 ${DEFAULT_CONTAINER_PADDING}`, ...style },
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "data-dt-component": dataDtComponent,
      "data-testid": dataTestId,
      "aria-label": ariaLabel,
      role,
      children: /* @__PURE__ */ jsxs(ChartLayout.Provider, { provider: PieChartSlicesDataProvider, children: [
        /* @__PURE__ */ jsx(ChartLayout.Graph, { children: /* @__PURE__ */ jsx(PieChartOverlays, { children: /* @__PURE__ */ jsx(PieChartPlot, { ...props, ref: mergedRef }) }) }),
        !legendHidden && /* @__PURE__ */ jsx(
          ChartLayout.Legend,
          {
            position: legendPosition,
            initialRatio: legendRatio,
            ratioBoundaries: PIE_CATEGORICAL_LEGEND_SIZES,
            overrideRatio,
            onResize: onRatioChange,
            children: /* @__PURE__ */ jsx(PieChartLegend, {})
          }
        )
      ] })
    }
  ) }) });
});
PieChartInternal.displayName = "PieChartInternal";
export {
  PieChartInternal
};
