// @dynatrace/strato-components/esm/charts/timeseries/constants.js
import { initialSelectionState } from "../core/components/annotations/providers/AnnotationsDataInitialState.js";
import { defaultRatio, SCALE_FALLBACK_COLOR } from "../core/constants.js";
import {
  GAP_POLICY_GAP
} from "../core/types/gap-policy.js";
const DISPLAY_NAME = "TimeseriesChart";
const defaultTooltipVariant = "single";
const defaultSeriesDisplayMode = "multi-line";
const defaultYAxis = {};
const defaultGapPolicy = GAP_POLICY_GAP;
const defaultSeriesCurve = "linear";
const defaultPointsDisplay = "auto";
const defaultValueRepresentation = "absolute";
const defaultInfiniteZoom = false;
const DEFAULT_SCALE_FALLBACK_COLOR = SCALE_FALLBACK_COLOR;
const defaultLegend = {
  hidden: false,
  position: "auto",
  ratio: defaultRatio
};
const defaultToolbar = {
  placement: "top-right",
  enabled: true
};
const defaultInteractions = {
  onZoomChange: void 0,
  pan: {
    enabled: true
  },
  zoom: {
    enabled: true
  }
};
const defaultColorPalette = "categorical";
const defaultTimeseriesChartConfig = {
  legend: defaultLegend,
  colorPalette: defaultColorPalette,
  gapPolicy: defaultGapPolicy,
  pointsDisplay: defaultPointsDisplay,
  valueRepresentation: defaultValueRepresentation,
  tooltip: {
    variant: defaultTooltipVariant,
    seriesDisplayMode: defaultSeriesDisplayMode
  },
  toolbar: defaultToolbar,
  interactions: defaultInteractions
};
const defaultTimeseriesPerVariant = {
  line: [],
  area: [],
  bar: [],
  band: [],
  shapePriority: []
};
const defaultThreshold = {
  left: [],
  right: []
};
const TIMESERIES_CHART_DEFAULT_VARIANT = "line";
const zIndicesTimeseriesAreas = {
  thresholdRange: 0,
  plot: 1,
  thresholdLine: 2,
  overlays: 3
};
const initialState = {
  filteredSeriesIds: /* @__PURE__ */ new Set(),
  highlightedSeriesId: void 0,
  annotations: initialSelectionState,
  zoom: {
    zoomInEnabled: true,
    zoomOutEnabled: false
  },
  currentDomain: void 0,
  select: void 0,
  interactionMode: "explore",
  explore: {
    visible: false,
    pinned: false,
    position: void 0
  },
  thresholds: { hoveredThresholdId: void 0 }
};
const TIMESERIES_ARIA_LABEL = "Timeseries Chart";
const TIMESERIES_INTERACTIONS_ARIA_LABEL = "Timeseries Chart interactions";
const TIMESERIES_TIME_AXIS_ARIA_LABEL = "timeseries-time-axis";
const AREA_CHART_DISPLAY_ARIA_LABEL = "area-chart-display";
const LINE_CHART_DISPLAY_ARIA_LABEL = "line-chart-display";
const BAR_CHART_DISPLAY_ARIA_LABEL = "bar-chart-display";
export {
  AREA_CHART_DISPLAY_ARIA_LABEL,
  BAR_CHART_DISPLAY_ARIA_LABEL,
  DEFAULT_SCALE_FALLBACK_COLOR,
  DISPLAY_NAME,
  LINE_CHART_DISPLAY_ARIA_LABEL,
  TIMESERIES_ARIA_LABEL,
  TIMESERIES_CHART_DEFAULT_VARIANT,
  TIMESERIES_INTERACTIONS_ARIA_LABEL,
  TIMESERIES_TIME_AXIS_ARIA_LABEL,
  defaultColorPalette,
  defaultGapPolicy,
  defaultInfiniteZoom,
  defaultInteractions,
  defaultLegend,
  defaultPointsDisplay,
  defaultSeriesCurve,
  defaultSeriesDisplayMode,
  defaultThreshold,
  defaultTimeseriesChartConfig,
  defaultTimeseriesPerVariant,
  defaultToolbar,
  defaultTooltipVariant,
  defaultValueRepresentation,
  defaultYAxis,
  initialState,
  zIndicesTimeseriesAreas
};
