// @dynatrace/strato-components/esm/charts/timeseries/TimeseriesChart.js
import { jsx } from "react/jsx-runtime";
import { forwardRef } from "react";
import { Annotations } from "./slots/annotations/Annotations.js";
import { Marker } from "./slots/annotations/Marker.js";
import { AnnotationsTooltip } from "./slots/annotations/Tooltip.js";
import { Track } from "./slots/annotations/Track.js";
import { TimeseriesChartDownloadCSVSlot } from "./slots/DownloadCSV.js";
import { Legend } from "./slots/Legend.js";
import { Pan } from "./slots/Pan.js";
import { Select } from "./slots/Select.js";
import { Area } from "./slots/shapes/Area.js";
import { Band } from "./slots/shapes/Band.js";
import { Bar } from "./slots/shapes/Bar.js";
import { Line } from "./slots/shapes/Line.js";
import { ThresholdIndicator } from "./slots/ThresholdIndicator.js";
import { Toolbar } from "./slots/Toolbar.js";
import { Tooltip } from "./slots/Tooltip.js";
import { XAxis } from "./slots/XAxis.js";
import { YAxis } from "./slots/YAxis.js";
import { Zoom } from "./slots/Zoom.js";
import { TimeseriesChartRoot } from "./TimeseriesChartRoot.js";
import { Intent } from "../../core/slots/Intent/Intent.js";
import { EmptyStateSlot } from "../core/slots/chart-states/EmptyStateSlot.js";
import { ErrorStateSlot } from "../core/slots/chart-states/ErrorStateSlot.js";
import { ColorRule } from "../core/slots/ColorRule.js";
import { isFeatureEnabled } from "../core/utils/is-feature-enabled.js";
import { TimeseriesToXY } from "../xy-chart-adoption/timeseries/TimeseriesToXY.js";
const TimeseriesChartSwitcher = forwardRef((props, forwardedRef) => {
  const { __useXYAdoption, ...rest } = props;
  if (__useXYAdoption ?? isFeatureEnabled("timeseriesToXY")) {
    return /* @__PURE__ */ jsx(TimeseriesToXY, { ...rest, ref: forwardedRef });
  }
  return /* @__PURE__ */ jsx(TimeseriesChartRoot, { ...rest, ref: forwardedRef });
});
TimeseriesChartSwitcher.displayName = "TimeseriesChart";
const TimeseriesChart = Object.assign(TimeseriesChartSwitcher, {
  Band,
  Legend,
  YAxis,
  XAxis,
  Tooltip,
  Annotations,
  Line,
  Area,
  Bar,
  ThresholdIndicator,
  Select,
  Intent,
  EmptyState: EmptyStateSlot,
  ErrorState: ErrorStateSlot,
  ColorRule,
  Toolbar,
  DownloadCSV: TimeseriesChartDownloadCSVSlot,
  Pan,
  Zoom
});
const TimeseriesAnnotations = Object.assign(
  {},
  {
    Marker,
    Track,
    Tooltip: AnnotationsTooltip
  }
);
export {
  TimeseriesAnnotations,
  TimeseriesChart,
  TimeseriesChartSwitcher
};
