// @dynatrace/strato-components/esm/charts/timeseries/TimeseriesChartRoot.js
import { jsx, jsxs } from "react/jsx-runtime";
import { forwardRef } from "react";
import { TimeseriesChartLegend } from "./components/legend/TimeseriesChartLegend.js";
import { TIMESERIES_CHART_DEFAULT_VARIANT } from "./constants.js";
import { TimeseriesRawDataContext } from "./context/TimeseriesRawData.context.js";
import { useTimeseriesChartConfig } from "./hooks/useTimeseriesChartConfig.js";
import { StoreProvider } from "./providers/StoreProvider.js";
import { TimeseriesChartIdProvider } from "./providers/TimeseriesChartIdProvider.js";
import { TimeseriesConfigProvider } from "./providers/TimeseriesConfigProviders.js";
import { TimeseriesDataProvider } from "./providers/TimeseriesDataProvider.js";
import { TimeseriesFilteredDataProvider } from "./providers/TimeseriesFilteredDataProvider.js";
import { TimeseriesMinMaxXAxisProvider } from "./providers/TimeseriesMinMaxXAxisProvider.js";
import { TimeseriesChartInternal } from "./TimeseriesChartInternal.js";
import {
  hasSeriesWithDatapoints,
  splitTimeseriesPerVariant
} from "./utils/timeseries.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { ChartLayout } from "../core/components/chart-layout/ChartLayout.js";
import { useAutoLegendRefresh } from "../core/components/chart-layout/hooks/useAutoLegendRefresh.js";
import { getChartStatesTemplates } from "../core/components/chart-states/utils/get-chart-states-template.js";
import { OverlayProvider } from "../core/components/overlay-tooltip/contexts/OverlayContext.js";
import { TooltipProvider } from "../core/components/overlay-tooltip/contexts/TooltipContext.js";
import { ChartSeriesActionContext } from "../core/components/series-action/ChartSeriesActionContext.js";
import { coerceSizeValue } from "../core/utils/coerceSizeValue.js";
const TIMESERIES_CHART_DEFAULT_HEIGHT = 300;
const COMPONENT_NAME_BEHAVIOURAL_TRACKING = "TimeseriesChart";
const TimeseriesChartRoot = forwardRef((props, forwardedRef) => {
  const {
    children,
    data,
    gapPolicy,
    pointsDisplay,
    colorPalette,
    truncationMode,
    variant = TIMESERIES_CHART_DEFAULT_VARIANT,
    valueRepresentation,
    height: heightProp,
    width,
    seriesActions,
    onMessage,
    loading = false,
    curve,
    initialZoom,
    currentZoom,
    onZoomChange,
    infiniteZoom,
    className,
    style,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow
  } = props;
  const [behaviorTrackingProps] = useBehavioralTrackingProps(props);
  const chartHeight = coerceSizeValue(heightProp) || TIMESERIES_CHART_DEFAULT_HEIGHT;
  const dataPerVariant = splitTimeseriesPerVariant(
    data ?? [],
    children,
    variant
  );
  const config = useTimeseriesChartConfig(children, {
    gapPolicy,
    pointsDisplay,
    valueRepresentation,
    colorPalette,
    truncationMode,
    curve,
    initialZoom,
    currentZoom,
    onZoomChange,
    infiniteZoom
  });
  const { emptyState, errorState } = getChartStatesTemplates(children);
  const { shapePriority, ...rawDataPerVariant } = dataPerVariant;
  const hasSeries = hasSeriesWithDatapoints(dataPerVariant);
  const {
    position: legendPosition,
    onRatioChange,
    // eslint-disable-next-line @typescript-eslint/no-deprecated
    ratio,
    initialRatio,
    hidden: isLegendHidden
  } = config.legend;
  const legendRatio = initialRatio ?? ratio;
  const chartLayoutRef = useAutoLegendRefresh(data);
  return /* @__PURE__ */ jsx(TimeseriesChartIdProvider, { children: /* @__PURE__ */ jsx(TimeseriesConfigProvider, { config, children: /* @__PURE__ */ jsx(TimeseriesRawDataContext.Provider, { value: rawDataPerVariant, children: /* @__PURE__ */ jsx(
    TimeseriesDataProvider,
    {
      data: dataPerVariant,
      valueRepresentation: config.valueRepresentation,
      onMessage,
      children: /* @__PURE__ */ jsx(TimeseriesMinMaxXAxisProvider, { children: /* @__PURE__ */ jsx(
        ChartSeriesActionContext.Provider,
        {
          value: {
            chartSeriesActions: seriesActions
          },
          children: /* @__PURE__ */ jsx(
            StoreProvider,
            {
              data: dataPerVariant,
              currentDomain: config.currentZoom,
              initialDomain: config.initialZoom,
              infiniteZoom: config.infiniteZoom,
              children: /* @__PURE__ */ jsx(TimeseriesFilteredDataProvider, { children: /* @__PURE__ */ jsx(OverlayProvider, { children: /* @__PURE__ */ jsx(TooltipProvider, { children: /* @__PURE__ */ jsxs(
                ChartLayout,
                {
                  ref: chartLayoutRef,
                  chartHeight,
                  chartWidth: width,
                  isEmpty: !hasSeries,
                  emptyState,
                  errorState,
                  showLoader: loading,
                  className,
                  style,
                  ...behaviorTrackingProps,
                  "data-dtrum-mask": dataDtrumMask,
                  "data-dtrum-allow": dataDtrumAllow,
                  "data-dt-component": COMPONENT_NAME_BEHAVIOURAL_TRACKING,
                  children: [
                    /* @__PURE__ */ jsx(ChartLayout.Graph, { children: /* @__PURE__ */ jsx(
                      TimeseriesChartInternal,
                      {
                        ref: forwardedRef,
                        config
                      }
                    ) }),
                    !isLegendHidden && /* @__PURE__ */ jsx(
                      ChartLayout.Legend,
                      {
                        position: legendPosition,
                        initialRatio: legendRatio,
                        onResize: onRatioChange,
                        children: /* @__PURE__ */ jsx(TimeseriesChartLegend, {})
                      }
                    )
                  ]
                }
              ) }) }) })
            }
          )
        }
      ) })
    }
  ) }) }) });
});
TimeseriesChartRoot.displayName = "TimeseriesChartRoot";
export {
  TimeseriesChartRoot
};
