// @dynatrace/strato-components/charts/timeseries/layout/TimeseriesHighlightedCanvasLayout.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TimeseriesHighlightedCanvasLayout_exports = {};
__export(TimeseriesHighlightedCanvasLayout_exports, {
  TimeseriesHighlightedCanvasLayout: () => TimeseriesHighlightedCanvasLayout
});
module.exports = __toCommonJS(TimeseriesHighlightedCanvasLayout_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_TimeseriesGenericCanvasLayout = require("./TimeseriesGenericCanvasLayout.js");
var import_apply_opacity_to_canvas = require("../../core/utils/apply-opacity-to-canvas.js");
var import_safe_get = require("../../core/utils/safe-get.js");
var import_selectors = require("../hooks/store/selectors.js");
var import_useTimeseriesFilteredData = require("../hooks/useTimeseriesFilteredData.js");
const TimeseriesHighlightedCanvasLayout = ({
  handleOpacity
}) => {
  const highlightedSeriesId = (0, import_selectors.useHighlightedSeriesIdState)();
  const { dataPerVariant, seriesMap } = (0, import_useTimeseriesFilteredData.useTimeseriesFilteredData)();
  handleOpacity((0, import_apply_opacity_to_canvas.applyOpacityToCanvas)(highlightedSeriesId));
  if (!highlightedSeriesId) {
    return null;
  }
  const filterByHighlightedId = (series) => series.filter((s) => (0, import_safe_get.safeGet)(seriesMap, s).id === highlightedSeriesId);
  const filteredDataPerVariant = {
    line: filterByHighlightedId(dataPerVariant.line),
    area: filterByHighlightedId(dataPerVariant.area),
    bar: filterByHighlightedId(dataPerVariant.bar),
    band: filterByHighlightedId(dataPerVariant.band),
    shapePriority: dataPerVariant.shapePriority
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TimeseriesGenericCanvasLayout.TimeseriesGenericCanvasLayout, { dataPerVariant: filteredDataPerVariant });
};
