// @dynatrace/strato-components/charts/timeseries/layout/TimeseriesGenericCanvasLayout.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TimeseriesGenericCanvasLayout_exports = {};
__export(TimeseriesGenericCanvasLayout_exports, {
  TimeseriesGenericCanvasLayout: () => TimeseriesGenericCanvasLayout
});
module.exports = __toCommonJS(TimeseriesGenericCanvasLayout_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_Canvas = require("../../core/components/canvas/Canvas.js");
var import_CanvasContext = require("../../core/components/canvas/CanvasContext.js");
var import_useGridPlotSize = require("../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_constants = require("../../core/components/data-point/constants.js");
var import_AreaChartDisplay = require("../components/displays/area/AreaChartDisplay.js");
var import_BandChartDisplay = require("../components/displays/band/BandChartDisplay.js");
var import_BarChartDisplay = require("../components/displays/bar/BarChartDisplay.js");
var import_LineChartDisplay = require("../components/displays/line/LineChartDisplay.js");
var import_TimeseriesOverlays_css = require("../components/overlays/TimeseriesOverlays.sty.js");
var import_useXScale = require("../hooks/useXScale.js");
const displays = {
  line: import_LineChartDisplay.LineChartDisplay,
  bar: import_BarChartDisplay.BarChartDisplay,
  area: import_AreaChartDisplay.AreaChartDisplay
};
const TimeseriesGenericCanvasLayout = (0, import_react.forwardRef)(({ dataPerVariant }, canvasRefs) => {
  const { currentScale: xScale } = (0, import_useXScale.useXScales)();
  const { width, height } = (0, import_useGridPlotSize.useGridPlotSize)();
  const handleRef = (el, index) => {
    if (canvasRefs && typeof canvasRefs !== "function" && canvasRefs.current && el) {
      canvasRefs.current[index] = el;
    }
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, { children: dataPerVariant.shapePriority.map((variant, index) => {
    let Display = import_LineChartDisplay.LineChartDisplay;
    if (variant !== "band") {
      Display = displays[variant];
    }
    const isDottedShape = variant === "line" || variant === "area";
    const canvasContainerClassName = isDottedShape ? import_TimeseriesOverlays_css.datapointCanvasPlot : import_TimeseriesOverlays_css.canvasPlot;
    const canvasWidth = width && isDottedShape ? width + import_constants.OVERLAY_PADDING : width;
    const canvasHeight = height && isDottedShape ? height + import_constants.OVERLAY_PADDING : height;
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "div",
      {
        className: canvasContainerClassName,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_CanvasContext.CanvasContext.Provider,
          {
            value: {
              shapeRefs: /* @__PURE__ */ new Map()
            },
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_Canvas.Canvas,
              {
                ref: (el) => handleRef(el, index),
                width: canvasWidth,
                height: canvasHeight,
                children: variant === "band" ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  import_BandChartDisplay.BandChartDisplay,
                  {
                    series: dataPerVariant.band,
                    xScale
                  }
                ) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  Display,
                  {
                    series: dataPerVariant[variant],
                    xScale
                  }
                )
              }
            )
          }
        )
      },
      `highlight-display-${variant}${index}`
    );
  }) });
});
