// @dynatrace/strato-components/charts/meter-bar/multi-meter-bar/components/MultiMeterBarChartOverlay/MultiMeterBarChartOverlayWithTooltip.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MultiMeterBarChartOverlayWithTooltip_exports = {};
__export(MultiMeterBarChartOverlayWithTooltip_exports, {
  MultiMeterBarChartOverlayWithTooltip: () => MultiMeterBarChartOverlayWithTooltip
});
module.exports = __toCommonJS(MultiMeterBarChartOverlayWithTooltip_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_Flex = require("../../../../../layouts/flex/Flex.js");
var import_InteractionsOverlay_css = require("../../../../core/components/interactions-overlay/InteractionsOverlay.sty.js");
var import_utils = require("../../../../core/components/overlay-tooltip/_to-be-removed/utils.js");
var import_OverlayContext = require("../../../../core/components/overlay-tooltip/contexts/OverlayContext.js");
var import_OverlayTooltip = require("../../../../core/components/overlay-tooltip/OverlayTooltip.js");
var import_utils2 = require("../../../../core/components/overlay-tooltip/utils.js");
var import_ChartSeriesActionContext = require("../../../../core/components/series-action/ChartSeriesActionContext.js");
var import_MeterBarChartTooltip = require("../../../shared/components/meter-bar-tooltip/MeterBarChartTooltip.js");
var import_FormatterContext = require("../../../shared/context/FormatterContext.js");
var import_MeterBarLayoutSizeContext = require("../../../shared/context/MeterBarLayoutSizeContext.js");
var import_element_size_css = require("../../../shared/styles/element-size.sty.js");
var import_styles_css = require("../../../shared/styles.sty.js");
var import_get_numeric_value = require("../../../shared/utils/get-numeric-value.js");
var import_constants = require("../../constants.js");
var import_MultiMeterBarChartTooltipContext = require("../../context/MultiMeterBarChartTooltipContext.js");
var import_selectors = require("../../hooks/store/selectors.js");
var import_useMultiMeterBarChartTooltipHandlers = require("../../hooks/useMultiMeterBarChartTooltipHandlers.js");
const MultiMeterBarCustomTooltip = ({
  template,
  visibleSegments,
  hoveredIndex
}) => {
  const segments = visibleSegments.map((s) => ({
    name: s.name ?? "",
    value: (0, import_get_numeric_value.getNumericValue)(s.value, s.valueAccessor),
    color: s.color ?? ""
  }));
  const activeSegment = hoveredIndex !== void 0 && segments[hoveredIndex] ? { ...segments[hoveredIndex], index: hoveredIndex } : void 0;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayTooltip.OverlayTooltip, { legacyTemplate: template({ segments, activeSegment }) });
};
const MultiMeterBarChartOverlayWithTooltip = (props) => {
  const [hoveredSegmentIndex, setHoveredSegmentIndex] = (0, import_react.useState)(void 0);
  const overlay = (0, import_OverlayContext.useOverlay)();
  const { chartSeriesActions } = (0, import_ChartSeriesActionContext.useSeriesActions)();
  const { tooltipConfig } = (0, import_MultiMeterBarChartTooltipContext.useMultiMeterBarChartTooltipData)();
  const { size } = (0, import_MeterBarLayoutSizeContext.useMeterBarLayoutSize)();
  const { formatter, unit } = (0, import_FormatterContext.useFormatter)();
  const visibleSegments = (0, import_selectors.useMultiMeterBarVisibleSegmentsState)();
  (0, import_react.useEffect)(() => {
    setHoveredSegmentIndex(void 0);
  }, [visibleSegments.length]);
  const isSharedTooltip = tooltipConfig.variant === import_constants.TOOLTIP_VARIANT.Shared;
  const sharedData = {
    geometry: "multimeterbar",
    series: visibleSegments.map((item) => ({
      name: item.name ?? "",
      value: formatter(item.value, unit),
      color: item.color
    }))
  };
  const { onClick, onPointerMove, onPointerLeave } = (0, import_useMultiMeterBarChartTooltipHandlers.useMultiMeterBarChartTooltipHandlers)(sharedData, setHoveredSegmentIndex);
  const { data: tooltipData } = (0, import_utils.useTooltipData)();
  const tooltipContent = tooltipConfig.template && !(0, import_utils2.isTooltipDataEmpty)(tooltipData) ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    MultiMeterBarCustomTooltip,
    {
      template: tooltipConfig.template,
      visibleSegments,
      hoveredIndex: hoveredSegmentIndex
    }
  ) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_MeterBarChartTooltip.MeterBarChartTooltip,
    {
      variant: isSharedTooltip ? import_constants.TOOLTIP_VARIANT.Shared : import_constants.TOOLTIP_VARIANT.Single,
      activeItem: hoveredSegmentIndex ?? -1,
      actionsForItem: chartSeriesActions ? (item) => () => chartSeriesActions(item) : void 0
    }
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_Flex.Flex,
    {
      ref: overlay.overlayRef,
      className: (0, import_clsx.default)(import_styles_css.container, (0, import_element_size_css.elementSize)({ size }), import_InteractionsOverlay_css.interactionsOverlay),
      "data-testid": "overlay-wrapper",
      onPointerMove,
      onPointerLeave,
      onClick,
      children: [
        props.children,
        tooltipContent
      ]
    }
  );
};
