// @dynatrace/strato-components/esm/charts/xy-chart/components/displays/area-accessors.js
import { isNil } from "lodash-es";
import { getBandCenteredPoint } from "../../utils/get-band-centered-point.js";
import { getCenterDate } from "../../utils/get-center-date.js";
import {
  isCategoricalAreaScale,
  isNumericalAreaScale,
  isTimeAreaScale
} from "../../utils/type-guards.js";
function applyNumericalBarScaleTransformations(args, accessorCoord, baseline, relAccessor, accumulator, binDomain, isSeriesStacked) {
  const isNotBaseline = baseline !== void 0 && accessorCoord !== baseline;
  let { v0, v1 } = args;
  if (isNotBaseline && accumulator !== void 0) {
    v0 = !isNil(v0) ? v0 + accumulator : accumulator;
    v1 = !isNil(v1) ? v1 + accumulator : accumulator;
  }
  if (relAccessor && isNotBaseline && !isNil(binDomain) && isSeriesStacked) {
    return {
      v0: !isNil(v0) ? v0 / binDomain * 100 : v0,
      v1: !isNil(v1) ? v1 / binDomain * 100 : v1,
      scale: args.scale
    };
  }
  if (relAccessor) {
    return {
      v0: !isNil(v0) ? relAccessor(v0) : v0,
      v1: !isNil(v1) ? relAccessor(v1) : v1,
      scale: args.scale
    };
  }
  return { v0, v1, scale: args.scale };
}
const getPrimaryAxisAccessor = (accessorKeys) => {
  return (scale, relAccessor) => {
    return (datapoint) => {
      const v0 = datapoint[accessorKeys[0]];
      const v1 = datapoint[accessorKeys[1]];
      if (v0 === void 0) {
        throw Error("unknown Accessor value for Area series");
      }
      const args = { scale, v0, v1 };
      if (isNumericalAreaScale(args)) {
        return args.scale(relAccessor ? relAccessor(args.v0) : args.v0);
      }
      if (isTimeAreaScale(args)) {
        const value = getCenterDate(args.v0, args.v1);
        return args.scale(value);
      }
      if (isCategoricalAreaScale(args)) {
        return getBandCenteredPoint(args.scale, args.v0);
      }
      throw Error("unknown Accessor value for Area series");
    };
  };
};
const getGrowingAccessors = (accessorKeys, baselineCoord) => {
  const baselineAccessor = (scale, datapointsMetadata, relAccessor, isSeriesStacked) => {
    return (datapoint) => {
      const v0 = datapoint[accessorKeys[0]];
      const v1 = datapoint[accessorKeys[1]];
      const { accumulator, baseline, binDomain } = datapointsMetadata.get(datapoint) ?? {};
      let args = { scale, v0, v1 };
      if (isNumericalAreaScale(args)) {
        args = applyNumericalBarScaleTransformations(
          args,
          baselineCoord === "x" ? "y" : "x",
          baseline,
          relAccessor,
          accumulator,
          binDomain,
          isSeriesStacked
        );
      }
      if (isNumericalAreaScale(args)) {
        const value = args.v0 ?? 0;
        return args.scale(value);
      }
      if (isTimeAreaScale(args)) {
        return args.scale(args.scale.invert(0));
      }
      if (isCategoricalAreaScale(args)) {
        return args.scale.range()[1];
      }
      throw Error("unknown baseline accessor for Area series");
    };
  };
  const endpointAccessor = (scale, datapointsMetadata, relAccessor, isSeriesStacked) => {
    return (datapoint) => {
      const v0 = datapoint[accessorKeys[0]];
      const v1 = datapoint[accessorKeys[1]];
      const { accumulator, baseline, binDomain } = datapointsMetadata.get(datapoint) ?? {};
      if (v1 === void 0) {
        throw Error(
          `${accessorKeys[1]} cannot be undefined for the growing accessor in the AreaSeries`
        );
      }
      let args = { scale, v0, v1 };
      if (isNumericalAreaScale(args)) {
        args = applyNumericalBarScaleTransformations(
          args,
          baselineCoord === "x" ? "y" : "x",
          baseline,
          relAccessor,
          accumulator,
          binDomain,
          isSeriesStacked
        );
      }
      if (isTimeAreaScale(args) || isNumericalAreaScale(args)) {
        return args.scale(args.v1);
      }
      if (isCategoricalAreaScale(args)) {
        const catWidth = Math.floor(args.scale.bandwidth() / 2);
        return (args.scale(args.v1) ?? 0) + catWidth;
      }
      throw Error("unknown value accessor for Area series");
    };
  };
  return { baselineAccessor, endpointAccessor };
};
const xAccessor = (scale, relAccessor) => getPrimaryAxisAccessor(["x0", "x1"])(scale, relAccessor);
const { baselineAccessor: y0Accessor, endpointAccessor: y1Accessor } = getGrowingAccessors(["y0", "y1"], "x");
const yAccessor = (scale, relAccessor) => getPrimaryAxisAccessor(["y0", "y1"])(scale, relAccessor);
const { baselineAccessor: x0Accessor, endpointAccessor: x1Accessor } = getGrowingAccessors(["x0", "x1"], "y");
export {
  applyNumericalBarScaleTransformations,
  x0Accessor,
  x1Accessor,
  xAccessor,
  y0Accessor,
  y1Accessor,
  yAccessor
};
