// @dynatrace/strato-components/esm/charts/xy-chart/components/displays/band-accessors.js
import { getBandCenteredPoint } from "../../utils/get-band-centered-point.js";
import { getCenterDate } from "../../utils/get-center-date.js";
import {
  isCategoricalAreaScale,
  isNumericalAreaScale,
  isTimeAreaScale
} from "../../utils/type-guards.js";
const getBandValueAccessor = (accessorKey) => {
  return (scale, relAccessor) => {
    return (datapoint) => {
      const value = datapoint[accessorKey];
      if (value === void 0) {
        throw Error(
          `${accessorKey} cannot be undefined for band chart accessor`
        );
      }
      const args = { scale, v0: value, v1: void 0 };
      if (isNumericalAreaScale(args)) {
        return args.scale(relAccessor ? relAccessor(args.v0) : args.v0);
      }
      if (isTimeAreaScale(args)) {
        const isXAccessor = accessorKey === "x0" || accessorKey === "x1";
        const hasAllBandCoordinates = datapoint.x0 !== void 0 && datapoint.x1 !== void 0 && datapoint.y0 !== void 0 && datapoint.y1 !== void 0;
        if (hasAllBandCoordinates && isXAccessor) {
          const centeredTimestamp = getCenterDate(
            datapoint.x0,
            datapoint.x1
          );
          return args.scale(new Date(centeredTimestamp));
        }
        return args.scale(args.v0);
      }
      if (isCategoricalAreaScale(args)) {
        return getBandCenteredPoint(args.scale, args.v0);
      }
      throw Error(`unknown scale type for band accessor ${accessorKey}`);
    };
  };
};
const bandX0Accessor = (scale, relAccessor) => getBandValueAccessor("x0")(scale, relAccessor);
const bandX1Accessor = (scale, relAccessor) => getBandValueAccessor("x1")(scale, relAccessor);
const bandY0Accessor = (scale, relAccessor) => getBandValueAccessor("y0")(scale, relAccessor);
const bandY1Accessor = (scale, relAccessor) => getBandValueAccessor("y1")(scale, relAccessor);
function xMiddleBandAccessor(scale, relAccessor) {
  return (datapoint) => {
    const x0 = bandX0Accessor(scale, relAccessor)(datapoint);
    const x1 = bandX1Accessor(scale, relAccessor)(datapoint);
    return (x0 + x1) / 2;
  };
}
function yMiddleBandAccessor(scale, relAccessor) {
  return (datapoint) => {
    const y0 = bandY0Accessor(scale, relAccessor)(datapoint);
    const y1 = bandY1Accessor(scale, relAccessor)(datapoint);
    return (y0 + y1) / 2;
  };
}
export {
  bandX0Accessor,
  bandX1Accessor,
  bandY0Accessor,
  bandY1Accessor,
  xMiddleBandAccessor,
  yMiddleBandAccessor
};
