// @dynatrace/strato-components/esm/charts/xy-chart/components/displays/bar-accessors.js
import { isNil } from "lodash-es";
import {
  isCategoricalBarScale,
  isNumericalBarScale,
  isTimeBarScale
} from "../../utils/type-guards.js";
const applyNumericalBarScaleTransformations = (args, accessorCoord, baseline, relAccessor, accumulator, binDomain) => {
  const isNotBaseline = baseline !== void 0 && accessorCoord !== baseline;
  let { v0, v1 } = args;
  if (isNotBaseline && accumulator !== void 0) {
    v0 = !isNil(v0) ? v0 + accumulator : accumulator;
    v1 = !isNil(v1) ? v1 + accumulator : accumulator;
  }
  if (relAccessor && isNotBaseline && !isNil(binDomain)) {
    return {
      v0: !isNil(v0) ? v0 / binDomain * 100 : v0,
      v1: !isNil(v1) ? v1 / binDomain * 100 : v1,
      scale: args.scale
    };
  }
  if (relAccessor) {
    return {
      v0: !isNil(v0) ? relAccessor(v0) : v0,
      v1: !isNil(v1) ? relAccessor(v1) : v1,
      scale: args.scale
    };
  }
  return { v0, v1, scale: args.scale };
};
function calculateBarOrigin(args, accessorCoord) {
  const { scale, v0, v1 } = args;
  if (!isNil(v0) && !isNil(v1)) {
    return Math.min(scale(v0), scale(v1));
  }
  if (!isNil(v0)) {
    return scale(v0);
  }
  if (isNil(v1)) {
    return void 0;
  }
  if (typeof v1 === "number" && v1 < 0) {
    return accessorCoord === "x" ? scale(v1) : scale(0);
  }
  return typeof v1 === "number" ? Math.min(scale(0), scale(v1)) : Math.min(scale(scale.invert(scale.range()[0])), scale(v1));
}
function genericAccessor(scale, datapointsMetadata, accessorCoord, dimensionScale, seriesId, relAccessor) {
  return (datapoint) => {
    const v0 = accessorCoord === "x" ? datapoint.x0 : datapoint.y0;
    const v1 = accessorCoord === "x" ? datapoint.x1 : datapoint.y1;
    let args = { scale, v0, v1 };
    const { baseline, accumulator, binDomain } = datapointsMetadata.get(datapoint) ?? {};
    if (isCategoricalBarScale(args)) {
      if (!isNil(args.v0)) {
        if (dimensionScale && seriesId) {
          return (args.scale(args.v0) ?? 0) + (dimensionScale(seriesId) ?? 0);
        }
        return args.scale(args.v0) ?? 0;
      }
      if (!isNil(args.v1)) {
        if (accessorCoord === "x") {
          const [min] = args.scale.range();
          return min;
        } else {
          const catWidth = Math.floor(args.scale.bandwidth() / 2);
          return (args.scale(args.v1) ?? 0) + catWidth;
        }
      }
    }
    if (isNumericalBarScale(args)) {
      args = applyNumericalBarScaleTransformations(
        args,
        accessorCoord,
        baseline,
        relAccessor,
        accumulator,
        binDomain
      );
    }
    if (isNumericalBarScale(args) || isTimeBarScale(args)) {
      const barOrigin = calculateBarOrigin(args, accessorCoord);
      if (!isNil(barOrigin)) {
        return barOrigin;
      }
    }
    throw Error(`unknown ${accessorCoord}Accessor value for Bar`);
  };
}
function xAccessor(scale, datapointsMetadata, dimensionScale, seriesId, relAccessor) {
  return genericAccessor(
    scale,
    datapointsMetadata,
    "x",
    dimensionScale,
    seriesId,
    relAccessor
  );
}
function yAccessor(scale, datapointsMetadata, dimensionScale, seriesId, relAccessor) {
  return genericAccessor(
    scale,
    datapointsMetadata,
    "y",
    dimensionScale,
    seriesId,
    relAccessor
  );
}
const calculateBarDimension = (args) => {
  const { scale, v0, v1 } = args;
  if (!isNil(v0) && !isNil(v1)) {
    return Math.abs(scale(v1) - scale(v0));
  }
  if (!isNil(v1)) {
    const zero = typeof v1 === "number" ? scale(0) : scale(scale.invert(scale.range()[0]));
    return Math.abs(zero - scale(v1));
  }
  if (!isNil(v0)) {
    return scale(v0);
  }
  return void 0;
};
function spanAccessor(scale, datapointsMetadata, accessorCoord, dimensionScale, relAccessor) {
  return (datapoint) => {
    const v0 = accessorCoord === "x" ? datapoint.x0 : datapoint.y0;
    const v1 = accessorCoord === "x" ? datapoint.x1 : datapoint.y1;
    let args = { scale, v0, v1 };
    const { baseline, accumulator, binDomain } = datapointsMetadata.get(datapoint) ?? {};
    if (isCategoricalBarScale(args)) {
      if (!isNil(args.v0)) {
        if (dimensionScale) {
          return dimensionScale.bandwidth();
        }
        return args.scale.bandwidth();
      }
      if (!isNil(args.v1)) {
        if (accessorCoord === "x") {
          const catWidth = Math.floor(args.scale.bandwidth() / 2);
          return (args.scale(args.v1) ?? 0) + catWidth;
        } else {
          const [, min] = scale.range();
          const catWidth = Math.floor(args.scale.bandwidth() / 2);
          const barY = (args.scale(args.v1) ?? 0) + catWidth;
          return min - barY;
        }
      }
    }
    if (isNumericalBarScale(args)) {
      args = applyNumericalBarScaleTransformations(
        args,
        accessorCoord,
        baseline,
        relAccessor,
        accumulator,
        binDomain
      );
    }
    if (isNumericalBarScale(args) || isTimeBarScale(args)) {
      const span = calculateBarDimension(args);
      if (!isNil(span)) {
        return span;
      }
    }
    throw Error(
      `unknown ${accessorCoord === "x" ? "width" : "height"}Accessor value for Bar`
    );
  };
}
function widthAccessor(scale, datapointsMetadata, dimensionScale, relAccessor) {
  return spanAccessor(
    scale,
    datapointsMetadata,
    "x",
    dimensionScale,
    relAccessor
  );
}
function heightAccessor(scale, datapointsMetadata, dimensionScale, relAccessor) {
  return spanAccessor(
    scale,
    datapointsMetadata,
    "y",
    dimensionScale,
    relAccessor
  );
}
export {
  applyNumericalBarScaleTransformations,
  heightAccessor,
  widthAccessor,
  xAccessor,
  yAccessor
};
