// @dynatrace/strato-components/esm/charts/xy-chart/components/displays/DotDisplay.js
import { jsx } from "react/jsx-runtime";
import { xAccessor, yAccessor } from "./dot-accessors.js";
import { DotCanvas } from "../../../core/components/canvas/components/DotCanvas.js";
import { useGridPlotSize } from "../../../core/components/chart-grid/hooks/useGridPlotSize.js";
import { DEFAULT_FILL_OPACITY } from "../../../core/constants.js";
import { useXYChartData } from "../../hooks/useXYChartData.js";
import { getCurrentScale } from "../../utils/get-scale.js";
import { relAccessor } from "../../utils/value-representation.js";
function DotDisplay({
  axisIdToScales,
  series,
  metadata
}) {
  const { height, width } = useGridPlotSize();
  const { axisIdToScaleDomain } = useXYChartData();
  const xScale = getCurrentScale(axisIdToScales, series.xAxisId);
  const yScale = getCurrentScale(axisIdToScales, series.yAxisId);
  const { xAxis, yAxis } = metadata;
  const xRelAcc = relAccessor(xAxis, axisIdToScaleDomain);
  const yRelAcc = relAccessor(yAxis, axisIdToScaleDomain);
  const { color, shape } = metadata;
  return /* @__PURE__ */ jsx(
    DotCanvas,
    {
      data: series.data,
      canvasHeight: height,
      canvasWidth: width,
      color: () => color.Default,
      opacity: DEFAULT_FILL_OPACITY,
      x: xAccessor(xScale, xRelAcc),
      y: yAccessor(yScale, yRelAcc),
      shape,
      enableDotOverflow: true
    },
    series.id
  );
}
DotDisplay["displayName"] = "DotDisplay";
export {
  DotDisplay
};
