// @dynatrace/strato-components/esm/charts/xy-chart/components/displays/rect-accessors.js
import { safeGet } from "../../../core/utils/safe-get.js";
import { getColorFromScaleAndData } from "../../utils/coloring/get-color-from-scale-and-data.js";
import {
  isCategoricalRectScale,
  isNumericalRectScale,
  isTimeRectScale
} from "../../utils/type-guards.js";
function colorAccessor(colorScale, metaDataMap) {
  return (datapoint) => (
    // At this point we know value can not be null, undefined nor NaN, hence it will not be painted in the canvas
    getColorFromScaleAndData(
      colorScale,
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      datapoint.value,
      safeGet(metaDataMap, datapoint)?.colorOverride
    )
  );
}
function genericAccessor(scale, v0, v1, relAccessor) {
  const args = { scale, v0, v1 };
  if (isCategoricalRectScale(args)) {
    return args.scale(args.v0) ?? 0;
  } else if (isNumericalRectScale(args) || isTimeRectScale(args)) {
    const { scale: scale2 } = args;
    let { v0: v02, v1: v12 } = args;
    if (relAccessor) {
      if (isNumericalRectScale(args)) {
        v02 = relAccessor(v02);
        v12 = relAccessor(v12);
      }
    }
    return Math.min(scale2(v02), scale2(v12));
  }
  throw Error("unknown genericAccessor value for Rect");
}
function spanAccessor(scale, v0, v1, relAccessor) {
  const args = { scale, v0, v1 };
  if (isNumericalRectScale(args) || isTimeRectScale(args)) {
    const { scale: scale2 } = args;
    let { v0: v02, v1: v12 } = args;
    if (relAccessor) {
      if (isNumericalRectScale(args)) {
        v02 = relAccessor(v02);
        v12 = relAccessor(v12);
      }
    }
    return Math.abs(scale2(v12) - scale2(v02));
  } else if (isCategoricalRectScale(args)) {
    return args.scale.bandwidth();
  }
  throw Error("unknown spanAccessor value for Rect");
}
function xAccessor(scale, relAccessor) {
  return (datapoint) => genericAccessor(scale, datapoint.x0, datapoint.x1, relAccessor);
}
function yAccessor(scale, relAccessor) {
  return (datapoint) => genericAccessor(scale, datapoint.y0, datapoint.y1, relAccessor);
}
function widthAccessor(scale, relAccessor) {
  return (datapoint) => spanAccessor(scale, datapoint.x0, datapoint.x1, relAccessor);
}
function heightAccessor(scale, relAccessor) {
  return (datapoint) => spanAccessor(scale, datapoint.y0, datapoint.y1, relAccessor);
}
export {
  colorAccessor,
  heightAccessor,
  widthAccessor,
  xAccessor,
  yAccessor
};
