// @dynatrace/strato-components/esm/charts/xy-chart/components/displays/RectDisplay.js
import { jsx } from "react/jsx-runtime";
import {
  colorAccessor,
  heightAccessor,
  widthAccessor,
  xAccessor,
  yAccessor
} from "./rect-accessors.js";
import { BarCanvas } from "../../../core/components/canvas/components/BarCanvas.js";
import { useGridPlotSize } from "../../../core/components/chart-grid/hooks/useGridPlotSize.js";
import { DEFAULT_FILL_OPACITY } from "../../../core/constants.js";
import { safeGet } from "../../../core/utils/safe-get.js";
import { RECT_STROKE_WIDTH } from "../../constants/defaults.js";
import { useXYChartColorScales } from "../../hooks/useXYChartColorScales.js";
import { useXYChartData } from "../../hooks/useXYChartData.js";
import { getCurrentScale } from "../../utils/get-scale.js";
import { isGapDatapointForDisplay } from "../../utils/validation-datapoint.js";
import { relAccessor } from "../../utils/value-representation.js";
function RectDisplay(props) {
  const { axisIdToScales, series, rectFilterCallback } = props;
  const { height, width } = useGridPlotSize();
  const colorScalesMap = useXYChartColorScales();
  const { datapointMetadataMap, axisIdToScaleDomain, seriesToMetadata } = useXYChartData();
  return series.map((series2) => {
    const xScale = getCurrentScale(axisIdToScales, series2.xAxisId);
    const yScale = getCurrentScale(axisIdToScales, series2.yAxisId);
    const { xAxis, yAxis } = safeGet(seriesToMetadata, series2);
    const xRelAccessor = relAccessor(xAxis, axisIdToScaleDomain);
    const yRelAccessor = relAccessor(yAxis, axisIdToScaleDomain);
    const colorScale = colorScalesMap.get(series2.id);
    return colorScale ? /* @__PURE__ */ jsx(
      BarCanvas,
      {
        data: rectFilterCallback ? series2.filteredData.filter(rectFilterCallback) : series2.filteredData,
        canvasHeight: height,
        canvasWidth: width,
        color: colorAccessor(colorScale, datapointMetadataMap),
        x: xAccessor(xScale, xRelAccessor),
        width: widthAccessor(xScale, xRelAccessor),
        y: yAccessor(yScale, yRelAccessor),
        height: heightAccessor(yScale, yRelAccessor),
        opacity: () => DEFAULT_FILL_OPACITY,
        strokeWidth: RECT_STROKE_WIDTH,
        defined: isGapDatapointForDisplay(datapointMetadataMap)
      },
      series2.id
    ) : null;
  });
}
RectDisplay["displayName"] = "RectDisplay";
export {
  RectDisplay
};
