// @dynatrace/strato-components/charts/core/components/tooltip/shared/hooks/useChartTooltip.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useChartTooltip_exports = {};
__export(useChartTooltip_exports, {
  useChartTooltip: () => useChartTooltip
});
module.exports = __toCommonJS(useChartTooltip_exports);
var import_react = require("react");
var import_react_popper = require("react-popper");
var import_useOverlayStacking = require("../../../../../../core/hooks/useOverlayStacking.js");
var import_create_dom_rect = require("../../../../../../core/utils/create-dom-rect.js");
var import_popper_modifiers = require("../../../../../../overlays/utils/popper-modifiers.js");
function useChartTooltip(chartTooltipElement, defaultTooltipPlacement = "left") {
  const [virtualElement, setVirtualElement] = (0, import_react.useState)();
  const tooltipHeight = chartTooltipElement?.getBoundingClientRect().height;
  const popperModifiers = (0, import_react.useMemo)(() => {
    const modifiers = [
      {
        name: "flip",
        options: {
          fallbackPlacements: [defaultTooltipPlacement]
        }
      },
      {
        name: "preventOverflow",
        options: {
          mainAxis: true
          // true is the default value so maybe it can be removed
        }
      },
      {
        name: "computeStyles",
        options: {
          gpuAcceleration: false
        }
      },
      { name: "eventListeners", enabled: false },
      import_popper_modifiers.observeOverlaySize
    ];
    if (tooltipHeight) {
      modifiers.push({
        name: "offset",
        options: {
          offset: () => [
            // 24 Is the SizeLarge token, but since this is needed as a numeric, it is hardcoded here for now.
            Math.round(tooltipHeight / 2 - 24),
            // 16 Is the SizeLarge token, but since this is needed as a numeric, it is hardcoded here for now.
            16
          ]
          // [Aligns the tooltipmarker to the mouse, y-axis margin]
        }
      });
    }
    return modifiers;
  }, [tooltipHeight, defaultTooltipPlacement]);
  const { state, styles } = (0, import_react_popper.usePopper)(virtualElement, chartTooltipElement, {
    strategy: "fixed",
    placement: defaultTooltipPlacement === "left" ? "right" : "left",
    modifiers: popperModifiers
  });
  const overlayStacking = (0, import_useOverlayStacking.useOverlayStacking)(0, "overlay");
  (0, import_react.useEffect)(() => {
    const { top, left, right } = styles.popper;
    const display = virtualElement !== void 0 ? "block" : "none";
    chartTooltipElement?.setAttribute(
      "style",
      `display: ${display}; position: fixed; top: ${top}; left: ${left}; right: ${right}; z-index: ${overlayStacking.style.zIndex}`
    );
  }, [
    virtualElement,
    styles.popper,
    chartTooltipElement,
    overlayStacking.style.zIndex
  ]);
  const updateTooltipPosition = (0, import_react.useCallback)((mousePosition) => {
    if (mousePosition) {
      setVirtualElement({
        getBoundingClientRect: (0, import_create_dom_rect.createDOMRect)({
          x: mousePosition.x,
          y: mousePosition.y
        })
      });
    } else {
      setVirtualElement(void 0);
    }
  }, []);
  return {
    placement: state?.placement ?? defaultTooltipPlacement,
    updateTooltipPosition
  };
}
