// @dynatrace/strato-components/esm/charts/xy-chart/processing-strategies/utils/gap/get-gap-policy-information.js
import {
  GAP_POLICY_CONNECT,
  GAP_POLICY_GAP
} from "../../../../core/types/gap-policy.js";
import {
  getNumericalThresholdValue,
  getThresholdValue
} from "../../../../core/utils/apply-gap-policy.js";
import {
  getDatapointIsGap
} from "../../../../core/utils/get-data-point-is-gap.js";
import { isCategoricalAxis, isTimeAxis } from "../../../types/guards.js";
function getAxisGapPolicyThreshold(axis) {
  if (isCategoricalAxis(axis)) {
    return Infinity;
  }
  if (isTimeAxis(axis)) {
    return axis.gapPolicy === GAP_POLICY_CONNECT ? Infinity : getThresholdValue(axis.gapPolicy ?? GAP_POLICY_GAP);
  }
  return axis.gapPolicy === GAP_POLICY_CONNECT || axis.gapPolicy === void 0 ? Infinity : getNumericalThresholdValue(axis.gapPolicy);
}
function getDatapointDimensionValues(datapoint, axis, accessorKeys) {
  if (isTimeAxis(axis)) {
    return {
      start: datapoint[accessorKeys[0]],
      end: datapoint[accessorKeys[1]]
    };
  }
  return {
    start: datapoint[accessorKeys[0]] ?? datapoint[accessorKeys[1]]
  };
}
function getXDatapointDimensionValues(datapoint, axis) {
  return getDatapointDimensionValues(datapoint, axis, ["x0", "x1"]);
}
function getYDatapointDimensionValues(datapoint, axis) {
  return getDatapointDimensionValues(datapoint, axis, ["y0", "y1"]);
}
function detectGapBetweenDatapoints(lastDatapoint, currentDatapoint, xAxis, yAxis) {
  const xGapPolicyThreshold = getAxisGapPolicyThreshold(xAxis);
  const yGapPolicyThreshold = getAxisGapPolicyThreshold(yAxis);
  let isGap = false;
  if (xGapPolicyThreshold !== Infinity) {
    const lastDatapointX = getXDatapointDimensionValues(lastDatapoint, xAxis);
    const currentDatapointX = getXDatapointDimensionValues(
      currentDatapoint,
      xAxis
    );
    isGap = getDatapointIsGap(
      lastDatapointX,
      currentDatapointX,
      xGapPolicyThreshold
    );
  }
  if (!isGap && yGapPolicyThreshold !== Infinity) {
    const lastDatapointY = getYDatapointDimensionValues(lastDatapoint, yAxis);
    const currentDatapointY = getYDatapointDimensionValues(
      currentDatapoint,
      yAxis
    );
    isGap = getDatapointIsGap(
      lastDatapointY,
      currentDatapointY,
      yGapPolicyThreshold
    );
  }
  return isGap;
}
export {
  detectGapBetweenDatapoints,
  getAxisGapPolicyThreshold,
  getXDatapointDimensionValues,
  getYDatapointDimensionValues
};
