// @dynatrace/strato-components/esm/charts/core/components/interactions-overlay/constants.js
var KeyboardControls = /* @__PURE__ */ ((KeyboardControls2) => {
  KeyboardControls2["LEFT_ARROW"] = "ArrowLeft";
  KeyboardControls2["RIGHT_ARROW"] = "ArrowRight";
  KeyboardControls2["UP_ARROW"] = "ArrowUp";
  KeyboardControls2["DOWN_ARROW"] = "ArrowDown";
  KeyboardControls2["ENTER"] = "Enter";
  KeyboardControls2["ESCAPE"] = "Escape";
  KeyboardControls2["CMD"] = "Meta";
  KeyboardControls2["CONTROL"] = "Control";
  KeyboardControls2["ALT"] = "Alt";
  KeyboardControls2["SPACE"] = " ";
  KeyboardControls2["KEYBOARD_EXPLORE_MODE_SHORTCUT"] = "e";
  KeyboardControls2["KEYBOARD_EXPLORE_MODE_SHORTCUT_ALTERNATIVE"] = "E";
  KeyboardControls2["KEYBOARD_PAN_MODE_SHORTCUT"] = "p";
  KeyboardControls2["KEYBOARD_PAN_MODE_SHORTCUT_ALTERNATIVE"] = "P";
  KeyboardControls2["KEYBOARD_ZOOM_MODE_SHORTCUT"] = "z";
  KeyboardControls2["KEYBOARD_ZOOM_MODE_SHORTCUT_ALTERNATIVE"] = "Z";
  KeyboardControls2["KEYBOARD_RESET_SHORTCUT"] = "r";
  KeyboardControls2["KEYBOARD_RESET_SHORTCUT_ALTERNATIVE"] = "R";
  return KeyboardControls2;
})(KeyboardControls || {});
const MIN_ZOOM_STEP = 10;
const ZOOM_FACTOR = 0.1;
const SMALL_STEP_FACTOR = 0.01;
const LARGE_STEP_FACTOR = 0.02;
const SELECTION_EDGE_THRESHOLD = 2;
const SELECTION_RIGHT_OVERFLOW = 1;
export {
  KeyboardControls,
  LARGE_STEP_FACTOR,
  MIN_ZOOM_STEP,
  SELECTION_EDGE_THRESHOLD,
  SELECTION_RIGHT_OVERFLOW,
  SMALL_STEP_FACTOR,
  ZOOM_FACTOR
};
