// @dynatrace/strato-components/esm/charts/core/components/interactions-overlay/InteractionsOverlay.js
import { jsx } from "react/jsx-runtime";
import classNames from "clsx";
import {
  forwardRef,
  useRef,
  useState
} from "react";
import { chartEventsWrapper } from "./InteractionsOverlay.sty.js";
import { getActiveMode } from "./utils/get-active-mode.js";
import { getExploreInteractions } from "./utils/get-explore-interations.js";
import { getGestures } from "./utils/get-gestures.js";
import { getInterfaceInteractions } from "./utils/get-interface-interactions.js";
import { getPanInteractions } from "./utils/get-pan-interactions.js";
import { getZoomInteractions } from "./utils/get-zoom-interactions.js";
import { handleKeyDown } from "./utils/handle-keydown.js";
import { isOwnEvent } from "./utils/is-own-event.js";
import { useMergeRefs } from "../../../../core/hooks/useMergeRefs.js";
import { usePointerEvents } from "../../hooks/usePointerEvents.js";
import { usePreventScrollOnWheel } from "../../hooks/usePreventScrollOnWheel.js";
const InteractionsOverlay = forwardRef((props, forwardedRef) => {
  const {
    style,
    className,
    mode,
    pinned,
    selectionArea,
    width,
    height,
    onModeChange,
    onFocus,
    onHover,
    onLeave,
    onPanStart,
    onPanChange,
    onPanEnd,
    onTogglePin,
    onReset,
    onSelectStart,
    onSelectChange,
    onSelectEnd,
    onZoomStart,
    onZoomChange,
    onZoomEnd,
    isZoomDisabled,
    isPanDisabled,
    isZoomXDisabled,
    children,
    ...restProps
  } = props;
  const zoomDisabled = selectionArea ? true : props.isZoomDisabled;
  const panDisabled = selectionArea ? true : props.isPanDisabled;
  const panDeltaRef = useRef(void 0);
  const lastMousePositionRef = useRef(void 0);
  const isPointerDownRef = useRef(false);
  const eventCancelled = useRef(false);
  const handlerRef = useRef(null);
  const [cursorMode, setCursorMode] = useState(
    mode
  );
  const htmlRef = useRef(null);
  usePreventScrollOnWheel(htmlRef);
  const rootRef = useMergeRefs([htmlRef, forwardedRef]);
  const overlappingInteractionsProps = {
    selectionArea,
    eventCancelled,
    onSelectEnd,
    lastMousePositionRef,
    pinned,
    onHover,
    onTogglePin,
    setCursorMode,
    height,
    onReset,
    width,
    onZoomChange,
    onPanChange,
    panDeltaRef,
    onPanStart,
    onPanEnd
  };
  const handleMouseLeave = () => {
    onLeave?.({ triggeredBy: "pointer" });
    isPointerDownRef.current = false;
    lastMousePositionRef.current = void 0;
  };
  const handleFocus = (event) => {
    if (handlerRef.current) {
      handlerRef.current();
      handlerRef.current = null;
    }
    onFocus?.(event);
  };
  const handleBlur = (event) => {
    const nextFocusElement = event.relatedTarget;
    if (!event.target?.contains(nextFocusElement)) {
      onLeave?.({ triggeredBy: "keyboard" });
    }
  };
  const handleWheel = (event) => {
    interactions[getActiveMode(mode, event, zoomDisabled)].onWheel(event);
  };
  const handleKeyboard = (event) => handleKeyDown({
    event,
    lastMousePositionRef,
    panDeltaRef,
    selectionArea,
    eventCancelled,
    onSelectEnd,
    onModeChange,
    interaction: interactions[getActiveMode(mode, event, zoomDisabled, isPanDisabled)]
  });
  const interactions = {
    explore: getExploreInteractions({
      ...overlappingInteractionsProps,
      onModeChange,
      onSelectChange,
      onSelectStart,
      isZoomXDisabled
    }),
    zoom: getZoomInteractions({
      ...overlappingInteractionsProps,
      cursorMode,
      onSelectChange,
      onSelectStart
    }),
    pan: getPanInteractions({
      ...overlappingInteractionsProps,
      onModeChange,
      cursorMode
    })
  };
  const {
    handleMouseDown,
    handleMouseMove,
    handleMouseUp,
    handleTouchStart,
    handleTouchMove,
    handleTouchEnd
  } = usePointerEvents({
    ...getInterfaceInteractions(
      interactions,
      mode,
      zoomDisabled,
      isZoomXDisabled,
      panDisabled
    ),
    ...getGestures({
      lastMousePositionRef,
      panDeltaRef,
      onZoomStart,
      onZoomChange,
      onZoomEnd,
      onPanStart,
      onPanChange,
      onPanEnd
    }),
    eventCancelled
  });
  const handleDocumentMouseUp = (e) => {
    const { clientX, clientY } = e;
    if (htmlRef?.current) {
      handleMouseUp({
        ...e,
        currentTarget: htmlRef.current,
        clientX,
        clientY
      });
      isPointerDownRef.current = false;
      lastMousePositionRef.current = void 0;
    }
  };
  const handleDocumentMouseMove = (e) => {
    const { clientX, clientY } = e;
    if (htmlRef?.current) {
      handleMouseMove({
        ...e,
        currentTarget: htmlRef.current,
        clientX,
        clientY
      });
    }
  };
  const registerDocumentListeners = () => {
    if (!handlerRef.current) {
      document.addEventListener("mouseup", handleDocumentMouseUp);
      document.addEventListener("pointermove", handleDocumentMouseMove);
      handlerRef.current = () => {
        document.removeEventListener("mouseup", handleDocumentMouseUp);
        document.removeEventListener("pointermove", handleDocumentMouseMove);
      };
    }
  };
  return /* @__PURE__ */ jsx(
    "div",
    {
      ref: rootRef,
      role: "presentation",
      tabIndex: 0,
      className: classNames(
        chartEventsWrapper({ mode: cursorMode }),
        className
      ),
      style,
      onMouseDown: (event) => {
        if (isOwnEvent(event)) {
          isPointerDownRef.current = true;
          handleMouseDown(event);
        }
      },
      onMouseMove: (event) => {
        if (isOwnEvent(event)) {
          handleMouseMove(event);
        }
      },
      onMouseUp: (event) => {
        if (isOwnEvent(event)) {
          handleMouseUp(event);
          isPointerDownRef.current = false;
        }
      },
      onMouseLeave: (e) => {
        if (isPointerDownRef.current && lastMousePositionRef.current && htmlRef.current) {
          registerDocumentListeners();
        } else {
          handleMouseLeave();
        }
      },
      onTouchStart: (event) => {
        if (isOwnEvent(event)) {
          handleTouchStart(event);
        }
      },
      onTouchMove: (event) => {
        if (isOwnEvent(event)) {
          handleTouchMove(event);
        }
      },
      onTouchEnd: (event) => {
        if (isOwnEvent(event)) {
          handleTouchEnd(event);
        }
      },
      onMouseEnter: () => handleFocus(),
      onFocus: handleFocus,
      onBlur: handleBlur,
      onKeyDown: handleKeyboard,
      onWheel: handleWheel,
      ...restProps,
      children
    }
  );
});
InteractionsOverlay["displayName"] = "InteractionsOverlay";
export {
  InteractionsOverlay
};
