// @dynatrace/strato-components/esm/charts/core/components/tooltip/section-components/VirtualizedTooltipBody/VirtualizedTooltipBody.js
import { jsx } from "react/jsx-runtime";
import { useVirtualizer } from "@tanstack/react-virtual";
import clsx from "clsx";
import { useRef, Children, useEffect } from "react";
import { itemsList, listContainer } from "./VirtualizedTooltipBody.sty.js";
import { useBodyHeightStyle } from "../../ChartTooltip/hooks/useBodyHeightStyle.js";
import { ESTIMATED_ITEM_SIZE } from "../../constants.js";
import { container } from "../../shared/styles/TooltipBody.sty.js";
const VirtualizedTooltipBody = (props) => {
  const {
    children,
    estimatedItemSize,
    style: consumerStyle,
    className: consumerClassname,
    onScrollContent
  } = props;
  const bodyItems = Children.toArray(children);
  const bodyItemsCount = bodyItems.length;
  const scrollRef = useRef(null);
  const { getVirtualItems, getTotalSize, measureElement } = useVirtualizer({
    count: bodyItemsCount,
    getScrollElement: () => scrollRef.current,
    estimateSize: (index) => {
      return estimatedItemSize?.(index) ?? ESTIMATED_ITEM_SIZE;
    }
  });
  const items = getVirtualItems();
  const totalSize = getTotalSize();
  const bodyHeightStyle = useBodyHeightStyle(totalSize, bodyItemsCount);
  useEffect(() => {
    if (!onScrollContent) {
      return;
    }
    const scrollContainer = scrollRef.current;
    scrollContainer?.addEventListener("scroll", onScrollContent);
    return () => {
      scrollContainer?.removeEventListener("scroll", onScrollContent);
    };
  }, [onScrollContent]);
  if (!bodyItemsCount) {
    return null;
  }
  return /* @__PURE__ */ jsx(
    "div",
    {
      className: clsx(container, consumerClassname),
      ref: scrollRef,
      style: { ...bodyHeightStyle, ...consumerStyle },
      children: /* @__PURE__ */ jsx("div", { style: { height: totalSize }, className: listContainer, children: /* @__PURE__ */ jsx(
        "ul",
        {
          className: itemsList,
          style: { transform: `translateY(${items[0]?.start ?? 0}px)` },
          "data-testid": "virtualized-items-list",
          children: items.map((item) => /* @__PURE__ */ jsx("li", { "data-index": item.index, ref: measureElement, children: bodyItems[item.index] }, item.key))
        }
      ) })
    }
  );
};
export {
  VirtualizedTooltipBody
};
