// @dynatrace/strato-components/esm/filters/filter-field/parser/getNumberParser.js
class NumberFormatParser {
  formatter;
  getIndex;
  groupDelimiter;
  decimalDelimiter;
  numeralRegExp;
  groupRegExp;
  decimalRegExp;
  constructor(locale) {
    this.formatter = new Intl.NumberFormat(locale);
    const parts = this.formatter.formatToParts(12345.6789);
    const groupPart = parts.find((d) => d.type === "group");
    const decimalPart = parts.find((d) => d.type === "decimal");
    const numerals = [
      ...new Intl.NumberFormat(locale, { useGrouping: false }).format(
        9876543210
      )
    ].reverse();
    const indexMap = new Map(numerals.map((d, i) => [d, i]));
    this.getIndex = (d) => indexMap.get(d);
    this.numeralRegExp = new RegExp(`[${numerals.join("")}]`, "g");
    if (groupPart) {
      this.groupRegExp = new RegExp(`[${groupPart.value}]`, "g");
      this.groupDelimiter = groupPart.value;
    }
    if (decimalPart) {
      this.decimalRegExp = new RegExp(`[${decimalPart.value}]`);
      this.decimalDelimiter = decimalPart.value;
    }
  }
  parse(value) {
    const invalidDelimiter = value.match(
      new RegExp(
        `(\\${this.groupDelimiter}\\d{0,2}($|\\${this.groupDelimiter}))|(\\${this.decimalDelimiter}$)`
      )
    );
    if (invalidDelimiter !== null) {
      return NaN;
    }
    let parsed = value.trim();
    if (this.groupRegExp) {
      parsed = parsed.replace(this.groupRegExp, "");
    }
    if (this.decimalRegExp) {
      parsed = parsed.replace(this.decimalRegExp, ".");
    }
    if (this.numeralRegExp) {
      parsed = parsed.replace(
        this.numeralRegExp,
        (d) => this.getIndex(d)?.toString() ?? ""
      );
    }
    return parsed ? +parsed : NaN;
  }
}
let numberParserInstance;
function getNumberParser() {
  if (numberParserInstance === void 0) {
    numberParserInstance = new NumberFormatParser("en");
  }
  return numberParserInstance;
}
export {
  NumberFormatParser,
  getNumberParser
};
