// @dynatrace/strato-components/esm/filters/filter-field/parser/parser-context.js
import { ContextTracker } from "@lezer/lr";
class LastAcceptedStatement {
  lastToken;
  pos;
  validExpression;
  constructor(lastToken, pos, validExpression) {
    this.lastToken = lastToken;
    this.pos = pos;
    this.validExpression = validExpression;
  }
}
const trackTokens = [
  "expression",
  "comparison",
  "ComparisonOperator",
  "InclusionOperator",
  "SearchOperator",
  "Identifier",
  "JsonPathBase",
  "PositionOperator",
  "ClosingBracket",
  '","'
];
const trackLastToken = new ContextTracker({
  start: new LastAcceptedStatement(null, null),
  reduce(context, term, stack) {
    const termName = stack.parser.getName(term);
    if (trackTokens.includes(termName)) {
      return new LastAcceptedStatement(termName, stack.pos);
    }
    return context;
  },
  shift(context, term, stack) {
    const termName = stack.parser.getName(term);
    if (context.lastToken === "ComparisonOperator" && termName === "Identifier") {
      return new LastAcceptedStatement(context.lastToken, context.pos, true);
    }
    if (context.lastToken === "ComparisonOperator" && termName === "String") {
      return new LastAcceptedStatement(termName, stack.pos);
    }
    if (trackTokens.includes(termName)) {
      return new LastAcceptedStatement(termName, stack.pos);
    }
    return context;
  },
  strict: false
});
export {
  trackLastToken
};
