// @dynatrace/strato-components/esm/charts/core/components/legend/range/constants.js
import Colors from "@dynatrace/strato-design-tokens/colors";
import Spacings from "@dynatrace/strato-design-tokens/spacings";
import Typography from "@dynatrace/strato-design-tokens/typography";
import { format } from "@dynatrace-sdk/units";
const getNumberFromString = (value) => {
  const extractedNumber = value.match(/\d+/);
  return extractedNumber ? +extractedNumber[0] : 0;
};
const LABEL_ALIGNMENT = {
  Start: "start",
  Center: "center",
  End: "end"
};
const TRUNCATION_MODE = {
  End: "end",
  Middle: "middle",
  Start: "start"
};
const DEFAULT_TRUNCATION_MODE = TRUNCATION_MODE.Middle;
const DEFAULT_FORMATTER = (value) => format(value);
const DEFAULT_ITEM_THICKNESS = 10;
const DEFAULT_SEGMENT_SIZE = 6;
const DEFAULT_SEGMENT_COLOR = Colors.Background.Field.Neutral.Disabled;
const DEFAULT_GAP_THICKNESS = 1;
const DEFAULT_LABEL_FONT_SIZE = Typography.Text.Small.Default.Size;
const DEFAULT_LABEL_FONT_COLOR = Colors.Text.Neutral.Subdued;
const LABEL_FONT_COLOR_HOVERED = Colors.Text.Neutral.Default;
const LABEL_FONT_WEIGHT = Typography.Text.Small.Default.Weight;
const HOVERED_LABEL_FONT_WEIGHT = Typography.Text.Small.Emphasized.Weight;
const GAP_BETWEEN_LABELS = 8;
const TEXT_ELLIPSIS = "...";
const STROKE_WIDTH = 1;
const MAX_LEGEND_HEIGHT = 360;
const MAX_LEGEND_WIDTH = 360;
const CSS_CONTAINER_PADDING = Spacings.Size8;
const LABEL_HOVERED_PADDING = getNumberFromString(Spacings.Size8);
const LABEL_GAP = getNumberFromString(Spacings.Size2);
const CONTAINER_PADDING = getNumberFromString(CSS_CONTAINER_PADDING);
const DEFAULT_CONTAINER_SIZE = DEFAULT_ITEM_THICKNESS + CONTAINER_PADDING * 2;
const RANGE_LEGEND_CLASS = "range-legend";
const OVERLAY_ADDITIONAL_SIZE = getNumberFromString(Spacings.Size8);
const OVERLAY_OFFSET = OVERLAY_ADDITIONAL_SIZE / 2;
export {
  CONTAINER_PADDING,
  CSS_CONTAINER_PADDING,
  DEFAULT_CONTAINER_SIZE,
  DEFAULT_FORMATTER,
  DEFAULT_GAP_THICKNESS,
  DEFAULT_ITEM_THICKNESS,
  DEFAULT_LABEL_FONT_COLOR,
  DEFAULT_LABEL_FONT_SIZE,
  DEFAULT_SEGMENT_COLOR,
  DEFAULT_SEGMENT_SIZE,
  DEFAULT_TRUNCATION_MODE,
  GAP_BETWEEN_LABELS,
  HOVERED_LABEL_FONT_WEIGHT,
  LABEL_ALIGNMENT,
  LABEL_FONT_COLOR_HOVERED,
  LABEL_FONT_WEIGHT,
  LABEL_GAP,
  LABEL_HOVERED_PADDING,
  MAX_LEGEND_HEIGHT,
  MAX_LEGEND_WIDTH,
  OVERLAY_ADDITIONAL_SIZE,
  OVERLAY_OFFSET,
  RANGE_LEGEND_CLASS,
  STROKE_WIDTH,
  TEXT_ELLIPSIS,
  TRUNCATION_MODE
};
