// @dynatrace/strato-components/esm/forms/spin-buttons/text-spin-button/hooks/useValueNavigation.js
import {
  useCallback,
  useEffect,
  useMemo,
  useState
} from "react";
import { constrainValue } from "../../helper.js";
import { TEXT_SPIN_BUTTON_DEFAULTS } from "../constants.js";
const useValueNavigation = (props) => {
  const {
    options,
    onArrowUp = (idx, e) => {
      e.preventDefault();
      return idx - TEXT_SPIN_BUTTON_DEFAULTS.SMALL_STEP;
    },
    onArrowDown = (idx, e) => {
      e.preventDefault();
      return idx + TEXT_SPIN_BUTTON_DEFAULTS.SMALL_STEP;
    },
    onPageUp = (idx, e) => {
      e.preventDefault();
      return idx - TEXT_SPIN_BUTTON_DEFAULTS.BIG_STEP;
    },
    onPageDown = (idx, e) => {
      e.preventDefault();
      return idx + TEXT_SPIN_BUTTON_DEFAULTS.BIG_STEP;
    },
    onHome = (_, e) => {
      e.preventDefault();
      return 0;
    },
    onEnd = (_, e) => {
      e.preventDefault();
      return options.length - 1;
    },
    disabled,
    readOnly,
    defaultValue = TEXT_SPIN_BUTTON_DEFAULTS.VALUE,
    value: valueProp,
    onChange,
    isCyclic
  } = props;
  const [value, setValue] = useState(valueProp ?? defaultValue);
  const handleValueNavigation = useCallback(
    (transform, event) => {
      const sanitized = (idx) => {
        const maxIndex = options.length - 1;
        return constrainValue(idx, 0, maxIndex, isCyclic) ?? 0;
      };
      const curIndex = value === null ? 0 : options.indexOf(value || valueProp || "");
      const newIndex = sanitized(transform(curIndex, event));
      const newValue = options[newIndex];
      if (value !== newValue) {
        setValue(newValue);
        onChange?.(newValue);
      }
    },
    [value, options, valueProp, onChange, isCyclic]
  );
  const lookupTable = useMemo(
    () => options.map((v) => v[0].toLowerCase()),
    [options]
  );
  const searchValue = useCallback(
    (key, fireOnChange = false) => {
      const idx = lookupTable.indexOf(key.toLowerCase());
      if (idx < 0) {
        return void 0;
      }
      const newValue = options[idx];
      if (value !== newValue) {
        setValue(newValue);
      }
      fireOnChange && onChange?.(newValue);
      return newValue;
    },
    [lookupTable, options, value, onChange]
  );
  const clearValue = useCallback(
    (fireOnChange = false) => {
      setValue(null);
      fireOnChange && onChange?.(null);
    },
    [onChange]
  );
  const onKeyDownHandler = useCallback(
    (e) => {
      if (disabled || readOnly) {
        return;
      }
      const navigationHandlers = {
        ArrowUp: onArrowUp,
        ArrowDown: onArrowDown,
        PageUp: onPageUp,
        PageDown: onPageDown,
        Home: onHome,
        End: onEnd
      };
      const navHandler = navigationHandlers[e.key];
      if (navHandler) {
        if (!e.shiftKey && e.key !== "Shift") {
          handleValueNavigation(navHandler, e);
        }
        return;
      }
      switch (e.key) {
        case "Backspace":
        case "Delete":
          if (value !== null) {
            setValue(null);
            onChange?.(null);
          }
          e.preventDefault();
          break;
        default:
          if (!e.metaKey && !e.ctrlKey && e.key.match(/^\w$/)) {
            const newValue = searchValue(e.key);
            if (newValue !== void 0 && newValue !== value) {
              onChange?.(newValue);
            }
            e.preventDefault();
          }
          break;
      }
    },
    [
      disabled,
      handleValueNavigation,
      onArrowDown,
      onArrowUp,
      onChange,
      onEnd,
      onHome,
      onPageDown,
      onPageUp,
      readOnly,
      searchValue,
      setValue,
      value
    ]
  );
  useEffect(() => {
    if (valueProp !== void 0) {
      setValue(valueProp);
    }
  }, [setValue, valueProp]);
  return { value, onKeyDownHandler, searchValue, clearValue };
};
export {
  useValueNavigation
};
