// @dynatrace/strato-components/filters/filter-field/components/FilterStatement.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var FilterStatement_exports = {};
__export(FilterStatement_exports, {
  FilterStatement: () => FilterStatement,
  NodeRenderer: () => NodeRenderer
});
module.exports = __toCommonJS(FilterStatement_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_CSSHighlight = require("./CSSHighlight.js");
var filterfieldStyles = __toESM(require("../FilterField.sty.js"));
var styles = __toESM(require("./FilterStatement.sty.js"));
var import_string_tree_conversion = require("../utils/conversion/string-tree-conversion.js");
var import_tree_node_guards = require("../utils/tree-node-guards.js");
function WhitespaceRenderer(props) {
  const { start: startNode, end: endNode } = props;
  if (!startNode || !endNode) {
    return null;
  }
  const numberOfSpaces = Math.max(
    0,
    (endNode.range.from ?? -1) - (startNode.range.to ?? -1)
  );
  return Array(numberOfSpaces).fill(" ").map((item, index) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { className: styles.whitespace, children: " " }, `${item}-${index}`));
}
function withNodeRenderer(className) {
  return function Component(props) {
    const { node } = props;
    const ref = (0, import_react.useRef)(null);
    const highlight = (0, import_react.useMemo)(
      () => (0, import_CSSHighlight.getOrCreateNamedHighlight)("filter-field-escape-characters"),
      []
    );
    (0, import_react.useEffect)(() => {
      const ranges = [];
      if (node?.textValue.includes("\\") && ref.current) {
        const textNode = ref.current.childNodes.item(0);
        if (textNode) {
          const textContent = textNode.textContent ?? "";
          let index = 0;
          while ((index = textContent.indexOf("\\", index)) !== -1) {
            if (index + 1 >= textContent.length) {
              break;
            }
            const range = new Range();
            range.setStart(textNode, index);
            range.setEnd(textNode, index + 2);
            ranges.push(range);
            index += 2;
            highlight.add(range);
          }
        }
      }
      return () => {
        for (const range of ranges) {
          highlight.delete(range);
        }
      };
    }, [highlight, node?.textValue]);
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { ref, className: (0, import_clsx.default)(styles.escapeCharacter, className), children: node?.textValue });
  };
}
const KeyRenderer = withNodeRenderer(
  filterfieldStyles.filterKey
);
const StringRenderer = withNodeRenderer(
  filterfieldStyles.filterValue
);
const ComparisonOperatorRenderer = withNodeRenderer(
  filterfieldStyles.operator
);
const ListRenderer = withNodeRenderer("");
const VariableRenderer = withNodeRenderer(
  filterfieldStyles.variable
);
const JSONPathRenderer = withNodeRenderer(
  filterfieldStyles.filterValue
);
const NumberRenderer = withNodeRenderer(
  filterfieldStyles.filterValue
);
const BigIntRenderer = withNodeRenderer(
  filterfieldStyles.filterValue
);
const BooleanRenderer = withNodeRenderer(
  filterfieldStyles.filterValue
);
const DurationRenderer = withNodeRenderer(
  filterfieldStyles.filterValue
);
const ExistsOperatorRenderer = withNodeRenderer(
  filterfieldStyles.operator
);
const SearchOperatorRenderer = withNodeRenderer(
  filterfieldStyles.operator
);
const InclusionOperatorRenderer = withNodeRenderer(
  filterfieldStyles.operator
);
const ErrorRenderer = withNodeRenderer(
  filterfieldStyles.error
);
function GroupRenderer(props) {
  const { node } = props;
  const lastChild = node.children[node.children.length - 1];
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("span", { children: [
    node.explicit ? "(" : "",
    node.children.map((childNode, index, allChildren) => {
      const isLast = index === allChildren.length - 1;
      const nextChild = node.children[index + 1];
      let joiner = null;
      if (node.explicitLogicalOperator && !isLast && nextChild?.type !== "Error" && (!("value" in nextChild) || nextChild.value !== "")) {
        joiner = /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("span", { className: filterfieldStyles.operator, children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { className: styles.whitespace, children: " " }),
          node.logicalOperator,
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { className: styles.whitespace, children: " " })
        ] });
      } else if (node.explicitLogicalOperator === false && !isLast) {
        joiner = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          WhitespaceRenderer,
          {
            start: childNode,
            end: allChildren.at(index + 1)
          }
        );
      }
      return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_react.Fragment, { children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(NodeRenderer, { node: childNode }),
        joiner
      ] }, `${childNode.range.from}-${childNode.range.to}`);
    }),
    node.explicit && lastChild.type !== "Error" && (!("value" in lastChild) || lastChild.value !== "") ? ")" : ""
  ] });
}
function ValueRenderer(props) {
  const { node } = props;
  if ((0, import_tree_node_guards.isFilterFieldLeafNode)(node)) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(LeafNodeRenderer, { node });
  }
  if ((0, import_tree_node_guards.isFilterFieldListNode)(node)) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(ListRenderer, { node });
  }
  return null;
}
function StatementRenderer(props) {
  const { node } = props;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("span", { className: styles.statement, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(KeyRenderer, { node: node.key }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(WhitespaceRenderer, { start: node.key, end: node.comparisonOperator }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(LeafNodeRenderer, { node: node.comparisonOperator }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(WhitespaceRenderer, { start: node.comparisonOperator, end: node.value }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(ValueRenderer, { node: node.value })
  ] });
}
function LeafNodeRenderer(props) {
  const { node } = props;
  if ((0, import_tree_node_guards.isFilterFieldKeyNode)(node)) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(KeyRenderer, { node });
  }
  if ((0, import_tree_node_guards.isFilterFieldComparisonOperatorNode)(node)) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(ComparisonOperatorRenderer, { node });
  }
  if ((0, import_tree_node_guards.isFilterFieldStringNode)(node)) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(StringRenderer, { node });
  }
  if ((0, import_tree_node_guards.isFilterFieldVariableNode)(node)) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(VariableRenderer, { node });
  }
  if ((0, import_tree_node_guards.isFilterFieldNumberNode)(node)) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(NumberRenderer, { node });
  }
  if ((0, import_tree_node_guards.isFilterFieldJSONPathNode)(node)) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(JSONPathRenderer, { node });
  }
  if ((0, import_tree_node_guards.isFilterFieldBigIntNode)(node)) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(BigIntRenderer, { node });
  }
  if ((0, import_tree_node_guards.isFilterFieldBooleanNode)(node)) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(BooleanRenderer, { node });
  }
  if ((0, import_tree_node_guards.isFilterFieldDurationNode)(node)) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(DurationRenderer, { node });
  }
  if ((0, import_tree_node_guards.isFilterFieldExistsOperatorNode)(node)) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(ExistsOperatorRenderer, { node });
  }
  if ((0, import_tree_node_guards.isFilterFieldSearchOperatorNode)(node)) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(SearchOperatorRenderer, { node });
  }
  if ((0, import_tree_node_guards.isFilterFieldInclusionOperatorNode)(node)) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(InclusionOperatorRenderer, { node });
  }
  if ((0, import_tree_node_guards.isFilterFieldErrorNode)(node)) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(ErrorRenderer, { node });
  }
  return null;
}
function NodeRenderer(props) {
  const { node } = props;
  if ((0, import_tree_node_guards.isFilterFieldGroupNode)(node)) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(GroupRenderer, { node });
  }
  if ((0, import_tree_node_guards.isFilterFieldStatementNode)(node)) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(StatementRenderer, { node });
  }
  if ((0, import_tree_node_guards.isFilterFieldErrorNode)(node)) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(ErrorRenderer, { node });
  }
  return null;
}
function FilterStatement(props) {
  const {
    value: filterString,
    "data-testid": dataTestId,
    className: consumerClassName,
    style: consumerStyle,
    highlighting = "syntax"
  } = props;
  const tokenizedValue = (0, import_react.useMemo)(() => {
    return (0, import_string_tree_conversion.convertStringToFilterFieldTree)(filterString ?? "", {
      parserConfig: {
        jsonPathConversion: true,
        searchConversion: true
      }
    });
  }, [filterString]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      className: (0, import_clsx.default)(styles.root, consumerClassName, [
        highlighting === "syntax" && styles.highlighting.syntax,
        highlighting === "minimal" && styles.highlighting.minimal
      ]),
      style: consumerStyle,
      "data-testid": dataTestId,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(NodeRenderer, { node: tokenizedValue.tree })
    }
  );
}
