// @dynatrace/strato-components/filters/filter-field/components/FilterStatement.sty.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var FilterStatement_css_exports = {};
__export(FilterStatement_css_exports, {
  escapeCharacter: () => escapeCharacter,
  highlight: () => highlight,
  highlighting: () => highlighting,
  root: () => root,
  statement: () => statement,
  whitespace: () => whitespace
});
module.exports = __toCommonJS(FilterStatement_css_exports);
var import_virtualFocus_css_ts_vanilla = require("../../../core/styles/virtualFocus.css");
var import_AppRoot_css_ts_vanilla = require("../../../core/components/app-root/AppRoot.css");
var import_FilterField_css_ts_vanilla = require("../FilterField.css");
var import_SuggestionsOverlay_css_ts_vanilla = require("../suggestions/SuggestionsOverlay.css");
var import_Suggestion_css_ts_vanilla = require("../suggestions/Suggestion.css");
var import_FilterStatement_css_ts_vanilla = require("./FilterStatement.css");
var escapeCharacter = "_ia7amm4-3-6-0";
var highlight = "_ia7amm5-3-6-0";
var highlighting = { syntax: "_ia7amm0-3-6-0", minimal: "_ia7amm1-3-6-0" };
var root = "_ia7amm2-3-6-0 _96tj13c-3-6-0";
var statement = "_ia7amm3-3-6-0";
var whitespace = "_ia7amm6-3-6-0";
