// @dynatrace/strato-components/filters/filter-field/components/FilterFieldSuffixButtons.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var FilterFieldSuffixButtons_exports = {};
__export(FilterFieldSuffixButtons_exports, {
  FilterFieldSuffixButtons: () => FilterFieldSuffixButtons
});
module.exports = __toCommonJS(FilterFieldSuffixButtons_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_state = require("@codemirror/state");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var styles = __toESM(require("./FilterFieldSuffixButtons.sty.js"));
var import_Button = require("../../../buttons/button/Button.js");
var import_Divider = require("../../../layouts/divider/Divider.js");
var import_Flex = require("../../../layouts/flex/Flex.js");
var import_useErrorAwareViewRef = require("../hooks/useErrorAwareViewRef.js");
var import_useErrorLogsStore = require("../hooks/useErrorLogsStore.js");
var import_useFilterFieldContext = require("../hooks/useFilterFieldContext.js");
var import_messages = require("../messages.js");
var import_SuggestionPin = require("../suggestions/SuggestionPin.js");
function FilterFieldSuffixButtons(props) {
  const { value, disabled, ariaDisabled, showError, readOnly } = props;
  const { filterNamespace, filterFieldId, editorView } = (0, import_useFilterFieldContext.useFilterFieldContext)();
  const intl = (0, import_react_intl.useIntl)();
  const { viewDispatch } = (0, import_useErrorAwareViewRef.useErrorAwareViewRef)({ current: editorView });
  const setAction = (0, import_useErrorLogsStore.useErrorLogsStore)((state) => state.setAction);
  if (!value) {
    return null;
  }
  const handleClearButton = () => {
    if (editorView) {
      setAction({
        type: "clearFilter",
        data: {
          value
        }
      });
      viewDispatch({
        changes: {
          from: 0,
          to: editorView?.state.doc.length,
          insert: ""
        },
        annotations: import_state.Transaction.userEvent.of("clear")
      });
      editorView.focus();
    }
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { className: styles.suffix, gap: 2, children: [
    filterNamespace && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_SuggestionPin.SuggestionPin,
        {
          size: "condensed",
          className: styles.button,
          disabled: disabled || readOnly,
          "aria-disabled": ariaDisabled,
          filterString: value,
          placement: "top",
          fallbackPlacements: ["top", "right", "left"],
          id: `${filterFieldId}-pin-button`,
          showError
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_Divider.Divider,
        {
          orientation: "vertical",
          className: styles.divider,
          color: showError ? "critical" : "neutral"
        }
      )
    ] }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Button.Button,
      {
        size: "condensed",
        onClick: handleClearButton,
        "aria-label": intl.formatMessage(import_messages.filterFieldMessages.clearFilter),
        color: showError ? "critical" : "neutral",
        className: styles.button,
        disabled: disabled || readOnly,
        "aria-disabled": ariaDisabled,
        "data-dt-name": "ClearAll",
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.XmarkIcon, {}) })
      }
    )
  ] });
}
