// @dynatrace/strato-components/esm/charts/core/components/overlay-tooltip/components/header/Header.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { assignInlineVars } from "@vanilla-extract/dynamic";
import clsx from "clsx";
import {
  headerContainer,
  headerSubText,
  headerText,
  headerTextMultiLine,
  headerWidthVar,
  maxLinesVar
} from "./Header.sty.js";
import { getShowRangeDuration } from "./utils/get-show-range-duration.js";
import { isHeaderEmpty } from "./utils/type-guards.js";
import { defaultLabelFormatterStrategy } from "./utils/value-formatters.js";
import { Divider } from "../../../../../../layouts/divider/Divider.js";
import { Tooltip } from "../../../../../../overlays/tooltip/Tooltip.js";
import {
  DEFAULT_HEADER_LINES,
  DEFAULT_TRUNCATION_MODE,
  MAX_HEADER_LINES,
  TOOLTIP_WRAPPER_PADDING
} from "../../constants.js";
import { useExternalConfig } from "../../contexts/ExternalConfigContext.js";
import { useTooltipDimensions } from "../../contexts/TooltipDimensionsContext.js";
import { useDefaultConfig } from "../../hooks/useDefaultConfig.js";
import { useDefaultLabels } from "../../hooks/useDefaultLabels.js";
import { divider, TOOLTIP_MAX_WIDTH } from "../../tooltip.sty.js";
import { computeHeaderDisplay } from "./utils/compute-header-display.js";
import { computeSubheaderDisplay } from "./utils/compute-subheader-display.js";
import { getNonGeometryHeader } from "./utils/get-non-geometry-header.js";
const Header = (props) => {
  const {
    header1,
    header2,
    subheader,
    label,
    label2,
    maxLines,
    duration1,
    duration2,
    showRangeDuration2
  } = props;
  const labelFor = useDefaultLabels();
  const defaultConfig = useDefaultConfig();
  const defaultStrategy = defaultConfig?.headerFormatStrategy ?? defaultLabelFormatterStrategy;
  const { displayHeaderDivider, truncationMode = DEFAULT_TRUNCATION_MODE } = useExternalConfig();
  const containerWidth = useTooltipDimensions()?.width || TOOLTIP_MAX_WIDTH;
  const showedLines = Math.min(
    maxLines ?? DEFAULT_HEADER_LINES,
    MAX_HEADER_LINES
  );
  const isSingleLine = showedLines === 1;
  const isTruncationEnd = truncationMode === "end";
  const showRangeDuration1 = getShowRangeDuration(
    defaultConfig,
    props.showRangeDuration1
  );
  const computedShowDuration2 = showRangeDuration2 ?? false;
  const formatStrategy = defaultStrategy ?? getNonGeometryHeader(header1);
  const header2FormatStrategy = getNonGeometryHeader(header2);
  const availableWidth = containerWidth - TOOLTIP_WRAPPER_PADDING * 2;
  if (isHeaderEmpty(header1, header2, subheader)) {
    return null;
  }
  const sharedDisplayParams = {
    labelFor,
    isSingleLine,
    isTruncationEnd,
    availableWidth,
    showedLines,
    truncationMode,
    showRangeDuration: showRangeDuration1
  };
  const {
    content: headerContent,
    totalTextWidth,
    shownText,
    isTruncated,
    durationContent,
    durationSuffix
  } = computeHeaderDisplay({
    header: header1,
    label,
    duration: duration1,
    formatStrategy,
    ...sharedDisplayParams
  });
  const {
    content: header2Content,
    totalTextWidth: header2TotalTextWidth,
    shownText: shownHeader2Text,
    isTruncated: isHeader2Truncated,
    durationContent: duration2Content,
    durationSuffix: duration2Suffix
  } = computeHeaderDisplay({
    header: header2,
    label: label2,
    duration: duration2,
    formatStrategy: header2FormatStrategy,
    ...sharedDisplayParams,
    showRangeDuration: computedShowDuration2
  });
  const { subheaderContent, shownSubheaderText, isSubheaderTruncated } = computeSubheaderDisplay({
    subheader,
    formatStrategy,
    availableWidth,
    truncationMode
  });
  const headerClassName = !isSingleLine && isTruncationEnd ? clsx(headerText, headerTextMultiLine) : headerText;
  const shouldConstrainWidth = totalTextWidth > availableWidth || header2TotalTextWidth > availableWidth;
  const headerContainerStyle = shouldConstrainWidth ? assignInlineVars({ [headerWidthVar]: `${availableWidth}px` }) : {};
  const isHeader2Unique = header2Content + duration2Suffix !== headerContent + durationSuffix;
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsxs("div", { className: headerContainer, style: headerContainerStyle, children: [
      /* @__PURE__ */ jsx(Tooltip, { text: headerContent, disabled: !isTruncated, children: /* @__PURE__ */ jsxs(
        "p",
        {
          className: headerClassName,
          style: assignInlineVars({ [maxLinesVar]: String(showedLines) }),
          children: [
            shownText,
            durationContent && durationContent !== "0" && /* @__PURE__ */ jsx("span", { className: headerSubText, children: durationSuffix })
          ]
        }
      ) }),
      shownHeader2Text && isHeader2Unique && /* @__PURE__ */ jsx("div", { children: /* @__PURE__ */ jsx(Tooltip, { text: header2Content, disabled: !isHeader2Truncated, children: /* @__PURE__ */ jsxs(
        "p",
        {
          className: headerClassName,
          style: assignInlineVars({
            [maxLinesVar]: String(showedLines)
          }),
          children: [
            shownHeader2Text,
            duration2Content && duration2Content !== "0" && /* @__PURE__ */ jsx("span", { className: headerSubText, children: duration2Suffix })
          ]
        }
      ) }) }),
      shownSubheaderText && /* @__PURE__ */ jsx("div", { children: /* @__PURE__ */ jsx(Tooltip, { text: subheaderContent, disabled: !isSubheaderTruncated, children: /* @__PURE__ */ jsx("p", { className: headerSubText, children: shownSubheaderText }) }) })
    ] }),
    displayHeaderDivider && /* @__PURE__ */ jsx(Divider, { orientation: "horizontal", className: divider }, "divider")
  ] });
};
export {
  Header
};
