// @dynatrace/strato-components/esm/charts/timeseries/components/displays/shared/utils/datapoint-stacking.js
import { formatUnit } from "@dynatrace-sdk/units";
import { isUnit } from "../../../../../core/types/formatter.js";
function getUnitString(unit) {
  if (unit === void 0) {
    return "undefined";
  }
  if (unit === null) {
    return "null";
  }
  if (unit === "") {
    return "";
  }
  if (isUnit(unit)) {
    return formatUnit(unit);
  }
  return unit.toString();
}
function getStackedKeyByUnit(sign, datapoint, unit) {
  return Math.sign(sign).toString() + datapoint.start.getTime() + datapoint.end?.getTime() + getUnitString(unit);
}
function getCumulativeValue(mapStackedPoints, datapoint, unit, seriesId, value = 0, sign) {
  const key = getStackedKeyByUnit(sign, datapoint, unit);
  const stackedDataPoint = mapStackedPoints.get(key);
  let datapointsCumulative = 0;
  let baseCumulative = 0;
  let cumulativeValue = value;
  if (stackedDataPoint) {
    const {
      datapointsCumulative: prevDatapointsCumulative,
      baseCumulative: prevBaseCumulative,
      seriesId: prevSeriesId
    } = stackedDataPoint;
    datapointsCumulative = prevDatapointsCumulative ?? 0;
    baseCumulative = prevBaseCumulative ?? 0;
    const baseCumulativeHasToChange = prevSeriesId !== seriesId;
    if (baseCumulativeHasToChange) {
      baseCumulative = datapointsCumulative;
    }
    cumulativeValue = baseCumulative + value;
  }
  datapointsCumulative = stackedDataPoint?.seriesId !== seriesId ? cumulativeValue : Math.sign(sign) > 0 ? Math.max(cumulativeValue, datapointsCumulative) : Math.min(cumulativeValue, datapointsCumulative);
  return { cumulativeValue, datapointsCumulative, baseCumulative, key };
}
export {
  getCumulativeValue,
  getStackedKeyByUnit
};
