// @dynatrace/strato-components/charts/xy-chart/components/annotations/utils/get-indicators-from-annotations.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var get_indicators_from_annotations_exports = {};
__export(get_indicators_from_annotations_exports, {
  getIndicatorsFromAnnotations: () => getIndicatorsFromAnnotations
});
module.exports = __toCommonJS(get_indicators_from_annotations_exports);
var import_indicator_visibility_helpers = require("../../../../core/components/annotations/components/indicators/utils/indicator-visibility-helpers.js");
var import_constants = require("../../../../core/components/annotations/constants.js");
var import_get_min_max_as_number = require("../../../../core/components/annotations/utils/get-min-max-as-number.js");
var import_populate_missing_marker_props = require("../../../../core/components/annotations/utils/populate-missing-marker-props.js");
var import_populate_missing_track_props = require("../../../../core/components/annotations/utils/populate-missing-track-props.js");
var import_serialize_marker_range = require("../../../../core/components/annotations/utils/serialize-marker-range.js");
var import_get_datapoints = require("../../../processing-strategies/utils/access-datapoint/get-datapoints.js");
var import_evaluate_with_data = require("../../../utils/evaluate-with-data.js");
const serializeAnnotationsMarkerFromXYChartData = (data, markerConfig) => {
  const response = {
    ...data,
    start: (0, import_get_datapoints.getDatumValue)(data, markerConfig.startAccessor),
    end: (0, import_get_datapoints.getDatumValue)(data, markerConfig.endAccessor),
    symbol: (0, import_evaluate_with_data.evaluateOptionalAttribute)(markerConfig.symbol, data, void 0),
    hidden: (0, import_evaluate_with_data.evaluateOptionalAttribute)(markerConfig.hidden, data, false),
    priority: (0, import_evaluate_with_data.evaluateOptionalAttribute)(markerConfig.priority, data, 0),
    color: markerConfig.colorAccessor ? (0, import_get_datapoints.getDatumValue)(data, markerConfig.colorAccessor) : void 0,
    title: markerConfig.titleAccessor ? (0, import_get_datapoints.getDatumValue)(data, markerConfig.titleAccessor) : void 0,
    description: markerConfig.descriptionAccessor ? (0, import_get_datapoints.getDatumValue)(data, markerConfig.descriptionAccessor) : void 0,
    indicatorsDisplay: (0, import_evaluate_with_data.evaluateOptionalAttribute)(
      markerConfig.indicatorsDisplay,
      data,
      void 0
    )
  };
  return response;
};
const serializeAnnotationsTrackFromXYChartData = (data, trackConfig) => {
  return {
    hidden: (0, import_evaluate_with_data.evaluateOptionalAttribute)(trackConfig.hidden, data, false),
    symbol: (0, import_evaluate_with_data.evaluateOptionalAttribute)(trackConfig.symbol, data, void 0),
    color: trackConfig.colorAccessor ? (0, import_get_datapoints.getDatumValue)(data, trackConfig.colorAccessor) : void 0,
    label: trackConfig.labelAccessor ? (0, import_get_datapoints.getDatumValue)(data, trackConfig.labelAccessor) : void 0,
    indicatorsDisplay: (0, import_evaluate_with_data.evaluateOptionalAttribute)(
      trackConfig.indicatorsDisplay,
      data,
      "auto"
    ),
    markers: []
  };
};
const getIndicatorsFromAnnotations = (data, tracksConfig, markersConfig, selection) => {
  const alwaysVisible = [];
  const selectable = /* @__PURE__ */ new Map();
  const trackIdTroughAccesor = tracksConfig.trackIdAccessor;
  const hiddenTracks = /* @__PURE__ */ new Set();
  const annotationsByTrack = data.reduce(
    (trackMap, datum) => {
      const trackId = (0, import_get_datapoints.getDatumValue)(datum, trackIdTroughAccesor);
      const annotationsTrack = trackMap.has(
        trackId
      ) ? trackMap.get(trackId) : (0, import_populate_missing_track_props.populateMissingTrackProps)(
        serializeAnnotationsTrackFromXYChartData(datum, tracksConfig)
      );
      if (annotationsTrack?.hidden) {
        hiddenTracks.add(trackId);
      }
      const marker = serializeAnnotationsMarkerFromXYChartData(
        datum,
        markersConfig
      );
      const populatedMarker = (0, import_populate_missing_marker_props.populateMissingMarkerProps)(marker, annotationsTrack);
      annotationsTrack.markers.push(populatedMarker);
      trackMap.set(trackId, annotationsTrack);
      const { color: trackColor, indicatorsDisplay: trackIndicatorDisplay } = annotationsTrack;
      const {
        color: markerColor,
        start: markerStartRaw,
        end: markerEndRaw,
        indicatorsDisplay: markerIndicatorDisplay,
        priority
      } = populatedMarker;
      const { min: markerStart, max: markerEnd } = (0, import_get_min_max_as_number.getMinMaxAsNumber)(
        markerStartRaw,
        markerEndRaw
      );
      if (!(0, import_indicator_visibility_helpers.isAlwaysHidden)(markerIndicatorDisplay, trackIndicatorDisplay)) {
        const trackIndex = Array.from(trackMap.keys()).findIndex(
          (name) => name === trackId
        );
        const markerId = (0, import_serialize_marker_range.serializeMarkerRange)(
          { ...populatedMarker, start: markerStart, end: markerEnd },
          trackIndex
        );
        const color = markerColor ?? trackColor ?? import_constants.DEFAULT_MARKER_COLOR;
        const indicator = (0, import_indicator_visibility_helpers.buildIndicator)(
          color,
          selection,
          markerId,
          markerStart,
          markerEnd,
          priority
        );
        if ((0, import_indicator_visibility_helpers.isAlwaysShown)(markerIndicatorDisplay, trackIndicatorDisplay)) {
          alwaysVisible.push(indicator);
        } else {
          selectable.set(markerId, indicator);
        }
      }
      return trackMap;
    },
    /* @__PURE__ */ new Map()
  );
  return {
    indicators: { alwaysVisible, selectable },
    annotationsByTrack,
    hiddenTracks
  };
};
