// @dynatrace/strato-components/overlays/sheet/Sheet.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Sheet_exports = {};
__export(Sheet_exports, {
  Sheet: () => Sheet
});
module.exports = __toCommonJS(Sheet_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_framer_motion = require("framer-motion");
var import_react = require("react");
var import_animations_js = __toESM(require("@dynatrace/strato-design-tokens/animations-js"));
var import_sheet_container_utils = require("./sheet-container-utils.js");
var import_Sheet_css = require("./Sheet.sty.js");
var import_FocusScope = require("../../core/components/focus-scope/FocusScope.js");
var import_OverlayContainer = require("../../core/components/OverlayContainer.js");
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useContainerSize = require("../../core/hooks/useContainerSize.js");
var import_useId = require("../../core/hooks/useId.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_useOverlayStacking = require("../../core/hooks/useOverlayStacking.js");
var import_OverlayStackingProvider = require("../../core/providers/OverlayStackingProvider.js");
var import_attempt_focus = require("../../core/utils/focus-management/attempt-focus.js");
var import_get_first_focusable_child = require("../../core/utils/focus-management/get-first-focusable-child.js");
var import_framer_motion_easings = require("../../core/utils/framer-motion-easings.js");
var import_merge_props = require("../../core/utils/merge-props.js");
var import_Flex = require("../../layouts/flex/Flex.js");
var import_Heading = require("../../typography/heading/Heading.js");
var import_useDialog = require("../hooks/useDialog.js");
var import_useOverlay = require("../hooks/useOverlay.js");
const COMPONENT_NAME = "Sheet";
const getAnimationVariants = (topOffset) => ({
  backdrop: {
    enterBackdrop: {
      // For the opacity we need to use the default value here. Framer motion would complain
      // about too complex values, when transitioning from two custom property values within
      // an rgba value apparently.
      opacity: 1,
      transition: {
        duration: import_animations_js.default.FadeIn.Duration.Fast,
        ease: import_framer_motion_easings.fadeInEasing
      }
    },
    exitBackdrop: {
      opacity: 0,
      transition: {
        duration: import_animations_js.default.FadeOut.Duration.Fast,
        ease: import_framer_motion_easings.fadeOutEasing,
        delay: import_animations_js.default.FadeOut.Duration.Base
      }
    }
  },
  sheet: {
    enterSlide: {
      top: topOffset,
      opacity: 1,
      transition: {
        duration: import_animations_js.default.SlideIn.Duration.Slow,
        ease: import_framer_motion_easings.slideInEasing
      }
    },
    exitSlide: () => ({
      top: `calc(${topOffset}px + 10%)`,
      opacity: 0.3,
      transition: {
        duration: import_animations_js.default.SlideOut.Duration.Slow,
        ease: import_framer_motion_easings.slideOutEasing
      }
    })
  },
  content: {
    enterFade: {
      opacity: 1,
      transition: {
        duration: import_animations_js.default.SlideIn.Duration.Slow,
        ease: import_framer_motion_easings.fadeInEasing
      }
    },
    exitFade: {
      opacity: 0,
      transition: {
        duration: import_animations_js.default.SlideOut.Duration.Slow,
        ease: import_framer_motion_easings.fadeOutEasing
      }
    }
  }
});
const Sheet = /* @__PURE__ */ (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const {
      show = false,
      title,
      children,
      onDismiss,
      actions,
      topoffset = 28,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...remainingProps
    } = props;
    const { "aria-labelledby": ariaLabelledBy, ...ariaLabelingProps } = (0, import_useAriaLabelingProps.useAriaLabelingProps)(remainingProps);
    const ref = (0, import_react.useRef)(null);
    const animationVariants = getAnimationVariants(topoffset);
    const { dialogProps } = (0, import_useDialog.useDialog)(remainingProps);
    const { overlayProps } = (0, import_useOverlay.useOverlay)(
      {
        ...props,
        onClose: () => {
          onDismiss?.();
        },
        isOpen: show,
        isDismissible: false,
        isKeyboardDismissDisabled: false
      },
      ref
    );
    const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
    const childrenScopeRef = (0, import_react.useRef)(null);
    const actionScopeRef = (0, import_react.useRef)(null);
    const focusFirstChildElement = () => {
      if (childrenScopeRef.current && actionScopeRef.current) {
        let firstFocusableChild = (0, import_get_first_focusable_child.getFirstFocusableChild)(
          childrenScopeRef.current
        );
        if (!firstFocusableChild) {
          firstFocusableChild = (0, import_get_first_focusable_child.getFirstFocusableChild)(actionScopeRef.current);
        }
        firstFocusableChild && (0, import_attempt_focus.attemptFocus)(firstFocusableChild);
      }
    };
    const containerRef = (0, import_react.useRef)(null);
    const containerSize = (0, import_useContainerSize.useContainerSize)(containerRef);
    const mergedRef = (0, import_useMergeRefs.useMergeRefs)([containerRef, forwardedRef]);
    const breakpointStyle = (0, import_react.useMemo)(
      () => (0, import_sheet_container_utils.calculateBreakpointFromContainerSize)(containerSize),
      [containerSize]
    );
    const overlayContainerProps = (0, import_useOverlayStacking.useOverlayStacking)(0, "overlay");
    const headingId = (0, import_useId.useId)();
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_OverlayContainer.OverlayContainer,
      {
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        style: { ...overlayContainerProps.style, ...consumerStyle },
        className: consumerClassName,
        ...restProps,
        ...behavioralTrackingProps,
        "data-dt-component": COMPONENT_NAME,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_framer_motion.AnimatePresence, { children: show && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_framer_motion.motion.div,
          {
            variants: animationVariants.backdrop,
            initial: "exitBackdrop",
            animate: "enterBackdrop",
            exit: "exitBackdrop",
            className: import_Sheet_css.sheetWrapperCSS,
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              "div",
              {
                ref: mergedRef,
                onPointerDown: (event) => {
                  if (event.target === event.currentTarget) {
                    onDismiss?.();
                  }
                },
                className: import_Sheet_css.sheetBackdropCSS,
                children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  import_FocusScope.FocusScope,
                  {
                    contain: true,
                    children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                      import_framer_motion.motion.div,
                      {
                        ...(0, import_merge_props.mergeProps)(overlayProps, dialogProps, {
                          style: breakpointStyle
                          // eslint-disable-next-line @typescript-eslint/no-explicit-any
                        }),
                        className: import_Sheet_css.sheetCSS,
                        "data-ismodal": true,
                        variants: animationVariants.sheet,
                        initial: "exitSlide",
                        animate: "enterSlide",
                        exit: "exitSlide",
                        onAnimationComplete: () => {
                          show && focusFirstChildElement();
                          window.dispatchEvent(new Event("animationDone"));
                        },
                        ref,
                        "aria-labelledby": ariaLabelledBy ?? headingId,
                        ...ariaLabelingProps,
                        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                          import_framer_motion.motion.div,
                          {
                            className: import_Sheet_css.sheetScrollableContainerCSS,
                            style: { height: `calc(100% - ${topoffset}px)` },
                            variants: animationVariants.content,
                            initial: "exitFade",
                            animate: "enterFade",
                            exit: "exitFade",
                            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayStackingProvider.OverlayStackingProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { gap: 16, flexDirection: "column", height: "100%", children: [
                              (title || actions) && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
                                import_Flex.Flex,
                                {
                                  justifyContent: title ? "space-between" : "flex-end",
                                  alignItems: "flex-start",
                                  children: [
                                    title && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                                      import_Heading.Heading,
                                      {
                                        level: 5,
                                        as: "h2",
                                        id: headingId,
                                        className: import_Sheet_css.sheetHeadingCSS,
                                        children: title
                                      }
                                    ),
                                    actions && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                                      import_Flex.Flex,
                                      {
                                        flexWrap: "wrap",
                                        justifyContent: "flex-end",
                                        ref: actionScopeRef,
                                        children: actions
                                      }
                                    )
                                  ]
                                }
                              ),
                              /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                                "div",
                                {
                                  ref: childrenScopeRef,
                                  style: {
                                    position: "relative",
                                    height: "100%",
                                    minHeight: 0
                                  },
                                  children
                                }
                              )
                            ] }) })
                          }
                        )
                      }
                    )
                  }
                )
              }
            )
          }
        ) })
      }
    );
  }
);
Sheet.displayName = COMPONENT_NAME;
