// @dynatrace/strato-geo/map/components/MapContent.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MapContent_exports = {};
__export(MapContent_exports, {
  MapContent: () => MapContent
});
module.exports = __toCommonJS(MapContent_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_maplibre = require("@vis.gl/react-maplibre");
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_charts = require("@dynatrace/strato-components/charts");
var import_core = require("@dynatrace/strato-components/core");
var import_constants = require("../constants.js");
var import_store = require("../store/store.js");
var import_MapToolbar = require("./toolbar/MapToolbar.js");
var import_layer_ids_context = require("../contexts/layer-ids.context.js");
var import_map_data_bounding_box_context = require("../contexts/map-data-bounding-box.context.js");
var import_use_active_interaction = require("../hooks/use-active-interaction.js");
var import_use_hover_interaction = require("../hooks/use-hover-interaction.js");
var import_use_map_base_layer_features = require("../hooks/use-map-base-layer-features.js");
var import_use_map_config = require("../hooks/use-map-config.js");
var import_use_map_loading = require("../hooks/use-map-loading.js");
var import_use_map_performance = require("../hooks/use-map-performance.js");
var import_use_overlay_events = require("../hooks/use-overlay-events.js");
var import_use_tooltip_event_listeners = require("../hooks/use-tooltip-event-listeners.js");
var import_use_webgl_context_error = require("../hooks/use-webgl-context-error.js");
var import_default_style = require("../map-styles/default-style.js");
var import_imperative_handler_provider = require("../providers/imperative-handler.provider.js");
var import_get_color_from_token = require("../utils/get-color-from-token.js");
var import_get_data_layers_bounding_box = require("../utils/get-data-layers-bounding-box.js");
var import_to_mapbox_initial_view_state = require("../utils/to-mapbox-initial-view-state.js");
var import_use_set_map_initial_configuration = require("../utils/use-set-map-initial-configuration.js");
var import_zoom_state_utils = require("./toolbar/utils/zoom-state-utils.js");
const lineLayer = {
  id: import_constants.BASE_LAYER_LINE_ID,
  type: "line"
};
const fillLayer = {
  id: import_constants.BASE_LAYER_FILL_ID,
  type: "fill"
};
const MapSource = () => {
  const baseLayerFeatures = (0, import_use_map_base_layer_features.useMapBaseLayerFeatures)();
  return baseLayerFeatures && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_react_maplibre.Source, { id: "data", type: "geojson", data: baseLayerFeatures, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_react_maplibre.Layer,
      {
        ...fillLayer,
        paint: {
          "fill-color": (0, import_get_color_from_token.getColorFromToken)(import_constants.DEFAULT_COUNTRIES_FILL_COLOR)
        }
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_react_maplibre.Layer,
      {
        ...lineLayer,
        type: "line",
        filter: [
          "any",
          ["==", "region_type", "COUNTRY"],
          ["==", "region_type", "REGION"]
        ],
        paint: {
          "line-color": (0, import_get_color_from_token.getColorFromToken)(import_constants.DEFAULT_BOUNDARIES_BORDER_COLOR),
          "line-width": [
            "case",
            ["==", ["get", "region_type"], "REGION"],
            import_constants.REGION_BORDER_WIDTH,
            import_constants.COUNTRY_BORDER_WIDTH
          ]
        }
      }
    )
  ] });
};
const MapInteractions = ({
  children,
  onContextLostError
}) => {
  const [contextLostError, setContextLostError] = (0, import_react.useState)(null);
  (0, import_use_hover_interaction.useHoverInteraction)();
  (0, import_use_active_interaction.useActiveInteraction)();
  (0, import_use_webgl_context_error.useWebGLContextError)(() => {
    onContextLostError?.();
    setContextLostError(
      new Error(
        "Too many active maps. This is a browser limitation. Try reducing the number of map instances."
      )
    );
  });
  if (contextLostError) {
    throw contextLostError;
  }
  return children;
};
const MapContent = (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    initialViewState: initialViewStateProp,
    children,
    style,
    mapStyle = import_default_style.defaultStyle,
    onViewStateChange,
    onMapLoad,
    onContextLostError,
    ...remaining
  } = props;
  const loading = (0, import_use_map_loading.useMapLoading)();
  const setZoomState = (0, import_store.useSetStateZoom)();
  const [savedInitialViewState, setSavedInitialViewState] = (0, import_react.useState)();
  const mapContainerRef = (0, import_react.useRef)(null);
  const mergedRefs = (0, import_core.useMergeRefs)([
    mapContainerRef,
    forwardedRef
  ]);
  const baseFeatureCollection = (0, import_use_map_base_layer_features.useMapBaseLayerFeatures)();
  const dataLayersBBox = (0, import_get_data_layers_bounding_box.getDataLayersBoundingBox)(
    children,
    baseFeatureCollection?.features
  );
  (0, import_use_tooltip_event_listeners.useTooltipEventListeners)();
  const { toolbar: toolbarConfig, interactions: interactionsConfig } = (0, import_use_map_config.useMapConfig)();
  const layerIds = (0, import_react.useContext)(import_layer_ids_context.LayerIdsContext);
  const shouldRenderToolbar = !(0, import_lodash_es.isUndefined)(toolbarConfig) && !toolbarConfig.hidden || !(0, import_lodash_es.isUndefined)(interactionsConfig);
  const { height: graphHeight, width: graphWidth } = (0, import_charts._useGraphSize)();
  const {
    handleMouseEnter,
    handleMouseMove,
    handleMouseLeave,
    handleMouseClick,
    handleZoom,
    handleDrag
  } = (0, import_use_overlay_events.useOverlayEvents)();
  const initialViewState = initialViewStateProp ?? dataLayersBBox;
  const { onLoadHandler, onRenderHandler } = (0, import_use_map_performance.useMapPerformance)(mapContainerRef);
  const calculatedInitialViewState = (0, import_to_mapbox_initial_view_state.toMapBoxInitialViewState)(initialViewState);
  const { zoomToFitInProgress } = (0, import_store.useMapZoomState)();
  const handleZoomEnd = (event) => {
    if (savedInitialViewState) {
      const zoomState = (0, import_zoom_state_utils.generateNewZoomState)(
        event,
        savedInitialViewState,
        zoomToFitInProgress
      );
      setZoomState(zoomState);
    }
  };
  const handleDragEnd = (event) => {
    if (savedInitialViewState) {
      const zoomState = (0, import_zoom_state_utils.generateNewZoomState)(
        event,
        savedInitialViewState,
        zoomToFitInProgress
      );
      setZoomState(zoomState);
    }
  };
  const isZoomEnabled = toolbarConfig?.zoom?.enabled !== false;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      "data-testid": "map-container",
      ref: mergedRefs,
      onMouseEnter: handleMouseEnter,
      onMouseLeave: handleMouseLeave,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_react_maplibre.Map,
        {
          style: {
            width: graphWidth,
            height: `${graphHeight}px`,
            ...style
          },
          onMove: (evt) => {
            onViewStateChange?.(evt.viewState);
          },
          initialViewState: calculatedInitialViewState,
          mapStyle,
          dragRotate: false,
          fadeDuration: 0,
          renderWorldCopies: false,
          interactiveLayerIds: layerIds,
          maxBounds: [
            [import_constants.MIN_LONGITUDE, import_constants.MIN_LATITUDE],
            [import_constants.MAX_LONGITUDE, import_constants.MAX_LATITUDE]
          ],
          onMouseMove: (e) => handleMouseMove(e),
          onClick: (e) => handleMouseClick(e),
          onZoom: handleZoom,
          onZoomEnd: handleZoomEnd,
          onDrag: handleDrag,
          onDragEnd: handleDragEnd,
          boxZoom: isZoomEnabled,
          keyboard: isZoomEnabled,
          scrollZoom: isZoomEnabled,
          doubleClickZoom: isZoomEnabled,
          attributionControl: false,
          onRender: () => onRenderHandler && onRenderHandler(),
          onLoad: (e) => {
            onMapLoad();
            onLoadHandler && onLoadHandler();
            const initialCenter = e.target.getCenter();
            setSavedInitialViewState({
              latitude: initialCenter.lat,
              longitude: initialCenter.lng,
              zoom: e.target.getZoom()
            });
          },
          ref: (0, import_use_set_map_initial_configuration.useSetMapInitialConfiguration)(),
          interactive: true,
          ...remaining,
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_map_data_bounding_box_context.MapDataBoundingBoxContext.Provider, { value: dataLayersBBox, children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
            import_imperative_handler_provider.MapViewImperativeHandler,
            {
              forwardedRef,
              containerRef: mapContainerRef,
              children: [
                /* @__PURE__ */ (0, import_jsx_runtime.jsx)(MapSource, {}),
                loading ? null : /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
                  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(MapInteractions, { onContextLostError, children }),
                  shouldRenderToolbar ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_MapToolbar.MapToolbar, {}) : null
                ] })
              ]
            }
          ) })
        }
      )
    }
  );
});
MapContent.displayName = "MapContent";
