// @dynatrace/strato-components/filters/filter-field/utils/error-utils/error-message-helpers.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var error_message_helpers_exports = {};
__export(error_message_helpers_exports, {
  getCustomTypeLabel: () => getCustomTypeLabel,
  getErrorMessageForType: () => getErrorMessageForType,
  getTypeLabelKey: () => getTypeLabelKey,
  isCustomType: () => isCustomType
});
module.exports = __toCommonJS(error_message_helpers_exports);
const builtInTypeLabelKeys = {
  Duration: "Durations",
  Number: "Numbers",
  String: "Strings",
  IPAddress: "IPAddresses",
  UID: "UIDs",
  Timestamp: "Timestamps",
  SmartscapeId: "SmartscapeIds",
  Boolean: "Boolean"
};
function getTypeLabelKey(type) {
  return builtInTypeLabelKeys[type];
}
function getCustomTypeLabel(type, customTypes) {
  if (!customTypes) {
    return void 0;
  }
  const customType = customTypes[type];
  if (customType && typeof customType === "object" && customType.label) {
    return customType.label;
  }
  return void 0;
}
function isCustomType(type) {
  return !builtInTypeLabelKeys[type] && type !== "Any" && type !== "JSONPath";
}
function getErrorMessageForType(type) {
  switch (type) {
    case "Boolean":
      return "invalidValueErrorBoolean";
    case "Duration":
      return "invalidValueErrorDurations";
    case "Number":
      return "invalidValueErrorNumbers";
    case "IPAddress":
      return "invalidValueErrorIPAddresses";
    case "UID":
      return "invalidValueErrorUIDs";
    case "SmartscapeId":
      return "invalidValueErrorSmartscapeIds";
    case "Timestamp":
      return "invalidValueErrorTimestamps";
    case "String":
      return "invalidValueErrorStrings";
    case "Any":
    default:
      return "invalidValueErrorGeneric";
  }
}
