// @dynatrace/strato-components/esm/charts/core/components/overlay-tooltip/components/header/utils/format-date-range.js
import {
  isSameDay,
  isSameHour,
  isSameMinute,
  isSameMonth,
  isSameYear
} from "date-fns";
import { formatDate } from "@dynatrace-sdk/units";
import { formatDateTicks } from "../../../../../utils/format-date-ticks.js";
import {
  formatWeekday,
  formatYear
} from "../../../../../utils/format-dates.js";
const shouldIncludeSeconds = (date) => date.getSeconds() !== 0;
const createFullDateTimeFormatOptions = (includeSeconds) => ({
  day: "2-digit",
  month: "short",
  year: "numeric",
  hour: "2-digit",
  minute: "2-digit",
  second: includeSeconds ? "numeric" : void 0
});
const formatRange = (from, to, prefix) => {
  const base = `${from} - ${to}`;
  return prefix ? `${prefix}, ${base}` : base;
};
const formatTimeRange = (from, to) => {
  return formatRange(formatDateTicks(from), formatDateTicks(to));
};
const formatMinuteRange = (from, to) => {
  const commonDate = formatDate(from, {
    month: "short",
    day: "2-digit",
    year: "numeric"
  });
  const fromTime = formatDate(from, { hour: "2-digit", minute: "2-digit" });
  const toTime = formatDate(to, { hour: "2-digit", minute: "2-digit" });
  return formatRange(fromTime, toTime, commonDate);
};
const formatHourRange = (from, to) => {
  const commonDay = formatDate(from, {
    day: "2-digit",
    month: "short",
    year: "numeric"
  });
  const fromHour = formatDate(from, { hour: "2-digit" });
  const toHour = formatDate(to, { hour: "2-digit" });
  return formatRange(fromHour, toHour, commonDay);
};
const formatDayRange = (from, to) => {
  const commonMonth = formatDate(from, {
    month: "short",
    year: "numeric"
  });
  const fromDay = formatDate(from, { day: "2-digit" });
  const toDay = formatDate(to, { day: "2-digit" });
  return formatRange(fromDay, toDay, commonMonth);
};
const formatYearRange = (from, to) => {
  const commonYear = formatYear(from.getTime());
  const fromWeekday = formatWeekday(from.getTime());
  const toWeekday = formatWeekday(to.getTime());
  return formatRange(fromWeekday, toWeekday, commonYear);
};
const formatDateRange = (from, to) => {
  if (isSameMinute(from, to)) {
    return formatTimeRange(from, to);
  }
  if (isSameHour(from, to)) {
    return formatMinuteRange(from, to);
  }
  if (isSameDay(from, to)) {
    return formatHourRange(from, to);
  }
  if (isSameMonth(from, to)) {
    return formatDayRange(from, to);
  }
  if (isSameYear(from, to)) {
    return formatYearRange(from, to);
  }
  const fromFull = formatDate(
    from,
    createFullDateTimeFormatOptions(shouldIncludeSeconds(from))
  );
  const toFull = formatDate(
    to,
    createFullDateTimeFormatOptions(shouldIncludeSeconds(to))
  );
  return formatRange(fromFull, toFull);
};
export {
  formatDateRange
};
