// @dynatrace/strato-components/esm/charts/core/components/overlay-tooltip/components/header/utils/type-guards.js
import { isEmpty, isNil } from "lodash-es";
const isDateRange = (data) => {
  return data instanceof Array && data.length === 2 && data.every((item) => item instanceof Date);
};
const isNumericRange = (data) => {
  return data instanceof Array && data.length === 2 && data.every((item) => typeof item === "number");
};
const isStringRange = (data) => {
  return data instanceof Array && data.length === 2 && data.every((item) => typeof item === "string");
};
const isNonEmptyRange = (data) => {
  return data instanceof Array && data.length === 2 && data.every((item) => !isNil(item));
};
const isFromTo = (data) => {
  return data instanceof Array && data.length === 2 && data.every((item) => !isNil(item));
};
const isSingleValue = (data) => data instanceof Array && data.length === 1 && !isNil(data[0]) || !(data instanceof Array) && !isNil(data);
const isHeaderEmpty = (header1, header2, subheader) => (isNil(header1) || isEmpty(header1)) && (isNil(header2) || isEmpty(header2)) && (isNil(subheader) || isEmpty(subheader));
export {
  isDateRange,
  isFromTo,
  isHeaderEmpty,
  isNonEmptyRange,
  isNumericRange,
  isSingleValue,
  isStringRange
};
