// @dynatrace/strato-components/esm/charts/timeseries/components/legend/TimeseriesChartLegend.js
import { jsx } from "react/jsx-runtime";
import { memo, useMemo } from "react";
import Colors from "@dynatrace/strato-design-tokens/colors";
import {
  BandChartIcon,
  BarChartIcon,
  LineChartIcon,
  StackedAreaChartIcon
} from "@dynatrace/strato-icons";
import { useChartLayoutLoading } from "../../../core/components/chart-layout/hooks/useChartLayoutLoading.js";
import { useChartLayoutOrientation } from "../../../core/components/chart-layout/hooks/useChartLayoutOrientation.js";
import { getLegendOrientation } from "../../../core/components/chart-layout/utils/orientation.js";
import { reduceExploreCancelTooltip } from "../../../core/components/interactions-state-helpers/reducers/explore-reducers.js";
import { ChartLegend } from "../../../core/components/legend/categorical/ChartLegend.js";
import { useLegendSeriesActions } from "../../../core/components/legend/categorical/hooks/useLegendSeriesActions.js";
import { safeGet } from "../../../core/utils/safe-get.js";
import { singleLineName } from "../../../core/utils/single-line-name.js";
import { useSetState } from "../../hooks/store/selectors.js";
import { useTimeseriesData } from "../../hooks/useTimeseriesData.js";
import { useTruncationMode } from "../../hooks/useTruncationMode.js";
import { hasMultiGeometries } from "../../utils/timeseries.js";
const renderIcon = (variant) => {
  switch (variant) {
    case "bar":
      return /* @__PURE__ */ jsx(BarChartIcon, { size: "small" });
    case "area":
      return /* @__PURE__ */ jsx(StackedAreaChartIcon, { size: "small" });
    case "band":
      return /* @__PURE__ */ jsx(BandChartIcon, { size: "small" });
    case "line":
    default:
      return /* @__PURE__ */ jsx(LineChartIcon, { size: "small" });
  }
};
const createItemData = (series, seriesMap) => {
  const isMultiGeometries = hasMultiGeometries(series);
  const result = [];
  for (const variant of series.shapePriority) {
    const variantSeries = series[variant];
    for (let index = 0; index < variantSeries.length; index++) {
      const singleSeries = variantSeries[index];
      const seriesMetadata = safeGet(seriesMap, singleSeries);
      result.push({
        id: seriesMetadata.id ?? `item-${index}`,
        color: seriesMetadata.color ?? Colors.Text.Neutral.Default,
        icon: isMultiGeometries ? renderIcon(variant) : void 0,
        name: singleLineName(singleSeries.name) ?? seriesMetadata.id
      });
    }
  }
  return result;
};
const TimeseriesChartLegend = memo(() => {
  const { dataPerVariant, data, seriesMap } = useTimeseriesData();
  const truncationMode = useTruncationMode();
  const setState = useSetState();
  const { orientation: layoutOrientation } = useChartLayoutOrientation();
  const isLoading = useChartLayoutLoading();
  const legendOrientation = getLegendOrientation(layoutOrientation);
  const legendActions = useLegendSeriesActions(
    data,
    (s) => safeGet(seriesMap, s).id
  );
  const legendItems = useMemo(
    () => createItemData(dataPerVariant, seriesMap),
    [dataPerVariant, seriesMap]
  );
  const updateHighlightedSeries = (item) => {
    setState((prev) => ({
      ...prev,
      highlightedSeriesId: item?.id
    }));
  };
  const updateVisibleSeries = (visibleSeries) => {
    const visibleSeriesSet = new Set(visibleSeries.map((s) => s.id));
    setState(
      (prev) => reduceExploreCancelTooltip({
        ...prev,
        filteredSeriesIds: visibleSeriesSet
      })
    );
  };
  return /* @__PURE__ */ jsx(
    ChartLegend,
    {
      items: legendItems,
      orientation: legendOrientation,
      interactionsDisabled: isLoading,
      onHover: updateHighlightedSeries,
      onClick: updateVisibleSeries,
      onSeriesAction: legendActions,
      truncationMode
    }
  );
});
TimeseriesChartLegend["displayName"] = "TimeseriesChartLegend";
export {
  TimeseriesChartLegend,
  renderIcon
};
