// @dynatrace/strato-components/charts/pie/constants.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var constants_exports = {};
__export(constants_exports, {
  CHART_VARIANT: () => CHART_VARIANT,
  DEFAULT_SCALE_FALLBACK_COLOR: () => DEFAULT_SCALE_FALLBACK_COLOR,
  DONUT_CHART_ARIA_LABEL: () => DONUT_CHART_ARIA_LABEL,
  DONUT_CHART_DISPLAY_NAME: () => DONUT_CHART_DISPLAY_NAME,
  LABEL_TYPE_OPTION: () => LABEL_TYPE_OPTION,
  PAD_ANGLE: () => PAD_ANGLE,
  PIE_CHART_ARIA_LABEL: () => PIE_CHART_ARIA_LABEL,
  PIE_CHART_DISPLAY_NAME: () => PIE_CHART_DISPLAY_NAME,
  THRESHOLD_TYPE_OPTION: () => THRESHOLD_TYPE_OPTION,
  TRUNCATION_MODE: () => TRUNCATION_MODE
});
module.exports = __toCommonJS(constants_exports);
var import_constants = require("../core/constants.js");
const DONUT_CHART_DISPLAY_NAME = "DonutChart";
const PIE_CHART_DISPLAY_NAME = "PieChart";
const CHART_VARIANT = Object.freeze({
  Donut: "donut",
  Pie: "pie"
});
const THRESHOLD_TYPE_OPTION = Object.freeze(
  {
    Absolute: "absolute",
    NumberOfSlices: "number-of-slices",
    Relative: "relative"
  }
);
const TRUNCATION_MODE = Object.freeze({
  End: "end",
  Middle: "middle",
  Start: "start"
});
const LABEL_TYPE_OPTION = Object.freeze({
  Absolute: "absolute",
  Relative: "relative"
});
const DEFAULT_SCALE_FALLBACK_COLOR = import_constants.SCALE_FALLBACK_COLOR;
const DONUT_CHART_ARIA_LABEL = "Donut Chart";
const PIE_CHART_ARIA_LABEL = "Pie Chart";
const PAD_ANGLE = 0.01;
