// @dynatrace/strato-components/charts/pie/DonutChart.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DonutChart_exports = {};
__export(DonutChart_exports, {
  DonutChart: () => DonutChart
});
module.exports = __toCommonJS(DonutChart_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_PieChartInternal = require("./components/PieChartInternal.js");
var import_constants = require("./constants.js");
var import_InnerElementContext = require("./context/InnerElementContext.js");
var import_usePieChartConfig = require("./hooks/usePieChartConfig.js");
var import_ChartProviders = require("./providers/ChartProviders.js");
var import_Toolbar = require("./shared/slots/Toolbar.js");
var import_DonutChartDownloadCSV = require("./slots/DonutChartDownloadCSV.js");
var import_DonutChartGrouping = require("./slots/DonutChartGrouping.js");
var import_DonutChartInner = require("./slots/DonutChartInner.js");
var import_DonutChartLabels = require("./slots/DonutChartLabels.js");
var import_DonutChartLegend = require("./slots/DonutChartLegend.js");
var import_DonutChartTooltip = require("./slots/DonutChartTooltip.js");
var import_build_and_download_csv = require("./utils/build-and-download-csv.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_Intent = require("../../core/slots/Intent/Intent.js");
var import_defaultChartValues_css = require("../../core/styles/defaultChartValues.sty.js");
var import_get_chart_states_template = require("../core/components/chart-states/utils/get-chart-states-template.js");
var import_get_toolbar_config_without_intents = require("../core/components/toolbar/utils/get-toolbar-config-without-intents.js");
var import_EmptyStateSlot = require("../core/slots/chart-states/EmptyStateSlot.js");
var import_ErrorStateSlot = require("../core/slots/chart-states/ErrorStateSlot.js");
var import_ColorRule = require("../core/slots/ColorRule.js");
var import_chart_types = require("../core/types/chart-types.js");
const COMPONENT_NAME = "DonutChart";
const sanitizeToolbarIntents = (config) => {
  const toolbar = config.toolbar ? (0, import_get_toolbar_config_without_intents.getToolbarConfigWithoutIntents)(config.toolbar) : void 0;
  return {
    ...config,
    toolbar
  };
};
const DonutChartRoot = (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    className,
    style,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "data-testid": dataTestId = "donut-chart-root"
  } = props;
  const chartConfig = (0, import_usePieChartConfig.usePieChartConfig)(props);
  const { emptyState, errorState } = (0, import_get_chart_states_template.getChartStatesTemplates)(props.children);
  const isDataAvailable = !(0, import_lodash_es.isEmpty)(props.data);
  const containerRef = (0, import_react.useRef)(null);
  const rootRef = (0, import_useMergeRefs.useMergeRefs)([
    forwardedRef,
    containerRef
  ]);
  (0, import_react.useImperativeHandle)(
    forwardedRef,
    () => ({
      element: containerRef.current,
      getConfig: () => JSON.stringify({
        type: import_chart_types.ChartType.Donut,
        ...sanitizeToolbarIntents(chartConfig)
      }),
      downloadData: () => {
        (0, import_build_and_download_csv.buildAndDownloadCsv)(props.data);
      }
    })
  );
  const providersProps = {
    chartData: props.data,
    chartConfig
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_InnerElementContext.InnerElementContext.Provider,
    {
      value: {
        innerElement: chartConfig.innerElement
      },
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ChartProviders.ChartProviders, { ...providersProps, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_PieChartInternal.PieChartInternal,
        {
          ...props,
          ref: rootRef,
          variant: import_constants.CHART_VARIANT.Donut,
          isEmpty: !isDataAvailable,
          emptyState,
          errorState,
          className: (0, import_clsx.default)(import_defaultChartValues_css.defaultChartSize, className),
          style,
          "data-dtrum-mask": dataDtrumMask,
          "data-dtrum-allow": dataDtrumAllow,
          "data-testid": dataTestId,
          "data-dt-component": COMPONENT_NAME,
          "aria-label": import_constants.DONUT_CHART_ARIA_LABEL,
          role: "img"
        }
      ) })
    }
  );
});
DonutChartRoot.displayName = import_constants.DONUT_CHART_DISPLAY_NAME;
const DonutChart = Object.assign(DonutChartRoot, {
  Legend: import_DonutChartLegend.Legend,
  Grouping: import_DonutChartGrouping.Grouping,
  Labels: import_DonutChartLabels.Labels,
  Inner: import_DonutChartInner.DonutChartInner,
  Tooltip: import_DonutChartTooltip.Tooltip,
  Intent: import_Intent.Intent,
  Toolbar: import_Toolbar.Toolbar,
  EmptyState: import_EmptyStateSlot.EmptyStateSlot,
  ErrorState: import_ErrorStateSlot.ErrorStateSlot,
  ColorRule: import_ColorRule.ColorRule,
  DownloadCSV: import_DonutChartDownloadCSV.DonutChartDownloadCSVSlot
});
