// @dynatrace/strato-components/charts/pie/messages.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var messages_exports = {};
__export(messages_exports, {
  PIE_CHART_MESSAGES: () => PIE_CHART_MESSAGES
});
module.exports = __toCommonJS(messages_exports);
var import_react_intl = require("react-intl");
const PIE_CHART_MESSAGES = (0, import_react_intl.defineMessages)({
  downloadAsCsv: {
    id: "9JhiinIAmqWKBJae",
    defaultMessage: "Download as CSV",
    description: 'Text in a tooltip shown when hovering the "Download data" button in the pie chart toolbar'
  },
  invalidDataset: {
    id: "qoM4Je2O4h920xx0",
    defaultMessage: "No data can be displayed because the value is 0",
    description: "Text displayed when the pie chart dataset is invalid"
  }
});
