// @dynatrace/strato-components/charts/core/components/overlay-tooltip/components/header/Header.sty.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Header_css_exports = {};
__export(Header_css_exports, {
  headerContainer: () => headerContainer,
  headerSubText: () => headerSubText,
  headerText: () => headerText,
  headerTextMultiLine: () => headerTextMultiLine,
  headerWidthVar: () => headerWidthVar,
  maxLinesVar: () => maxLinesVar
});
module.exports = __toCommonJS(Header_css_exports);
var import_Header_css_ts_vanilla = require("./Header.css");
var headerContainer = "_1ldhlar2-3-6-0";
var headerSubText = "_1ldhlar5-3-6-0";
var headerText = "_1ldhlar3-3-6-0";
var headerTextMultiLine = "_1ldhlar4-3-6-0";
var headerWidthVar = "var(--_1ldhlar1-3-6-0)";
var maxLinesVar = "var(--_1ldhlar0-3-6-0)";
