// @dynatrace/strato-components/charts/core/components/overlay-tooltip/components/header/Header.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Header_exports = {};
__export(Header_exports, {
  Header: () => Header
});
module.exports = __toCommonJS(Header_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_dynamic = require("@vanilla-extract/dynamic");
var import_clsx = __toESM(require("clsx"));
var import_Header_css = require("./Header.sty.js");
var import_get_show_range_duration = require("./utils/get-show-range-duration.js");
var import_type_guards = require("./utils/type-guards.js");
var import_value_formatters = require("./utils/value-formatters.js");
var import_Divider = require("../../../../../../layouts/divider/Divider.js");
var import_Tooltip = require("../../../../../../overlays/tooltip/Tooltip.js");
var import_constants = require("../../constants.js");
var import_ExternalConfigContext = require("../../contexts/ExternalConfigContext.js");
var import_TooltipDimensionsContext = require("../../contexts/TooltipDimensionsContext.js");
var import_useDefaultConfig = require("../../hooks/useDefaultConfig.js");
var import_useDefaultLabels = require("../../hooks/useDefaultLabels.js");
var import_tooltip_css = require("../../tooltip.sty.js");
var import_compute_header_display = require("./utils/compute-header-display.js");
var import_compute_subheader_display = require("./utils/compute-subheader-display.js");
var import_get_non_geometry_header = require("./utils/get-non-geometry-header.js");
const Header = (props) => {
  const {
    header1,
    header2,
    subheader,
    label,
    label2,
    maxLines,
    duration1,
    duration2,
    showRangeDuration2
  } = props;
  const labelFor = (0, import_useDefaultLabels.useDefaultLabels)();
  const defaultConfig = (0, import_useDefaultConfig.useDefaultConfig)();
  const defaultStrategy = defaultConfig?.headerFormatStrategy ?? import_value_formatters.defaultLabelFormatterStrategy;
  const { displayHeaderDivider, truncationMode = import_constants.DEFAULT_TRUNCATION_MODE } = (0, import_ExternalConfigContext.useExternalConfig)();
  const containerWidth = (0, import_TooltipDimensionsContext.useTooltipDimensions)()?.width || import_tooltip_css.TOOLTIP_MAX_WIDTH;
  const showedLines = Math.min(
    maxLines ?? import_constants.DEFAULT_HEADER_LINES,
    import_constants.MAX_HEADER_LINES
  );
  const isSingleLine = showedLines === 1;
  const isTruncationEnd = truncationMode === "end";
  const showRangeDuration1 = (0, import_get_show_range_duration.getShowRangeDuration)(
    defaultConfig,
    props.showRangeDuration1
  );
  const computedShowDuration2 = showRangeDuration2 ?? false;
  const formatStrategy = defaultStrategy ?? (0, import_get_non_geometry_header.getNonGeometryHeader)(header1);
  const header2FormatStrategy = (0, import_get_non_geometry_header.getNonGeometryHeader)(header2);
  const availableWidth = containerWidth - import_constants.TOOLTIP_WRAPPER_PADDING * 2;
  if ((0, import_type_guards.isHeaderEmpty)(header1, header2, subheader)) {
    return null;
  }
  const sharedDisplayParams = {
    labelFor,
    isSingleLine,
    isTruncationEnd,
    availableWidth,
    showedLines,
    truncationMode,
    showRangeDuration: showRangeDuration1
  };
  const {
    content: headerContent,
    totalTextWidth,
    shownText,
    isTruncated,
    durationContent,
    durationSuffix
  } = (0, import_compute_header_display.computeHeaderDisplay)({
    header: header1,
    label,
    duration: duration1,
    formatStrategy,
    ...sharedDisplayParams
  });
  const {
    content: header2Content,
    totalTextWidth: header2TotalTextWidth,
    shownText: shownHeader2Text,
    isTruncated: isHeader2Truncated,
    durationContent: duration2Content,
    durationSuffix: duration2Suffix
  } = (0, import_compute_header_display.computeHeaderDisplay)({
    header: header2,
    label: label2,
    duration: duration2,
    formatStrategy: header2FormatStrategy,
    ...sharedDisplayParams,
    showRangeDuration: computedShowDuration2
  });
  const { subheaderContent, shownSubheaderText, isSubheaderTruncated } = (0, import_compute_subheader_display.computeSubheaderDisplay)({
    subheader,
    formatStrategy,
    availableWidth,
    truncationMode
  });
  const headerClassName = !isSingleLine && isTruncationEnd ? (0, import_clsx.default)(import_Header_css.headerText, import_Header_css.headerTextMultiLine) : import_Header_css.headerText;
  const shouldConstrainWidth = totalTextWidth > availableWidth || header2TotalTextWidth > availableWidth;
  const headerContainerStyle = shouldConstrainWidth ? (0, import_dynamic.assignInlineVars)({ [import_Header_css.headerWidthVar]: `${availableWidth}px` }) : {};
  const isHeader2Unique = header2Content + duration2Suffix !== headerContent + durationSuffix;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { className: import_Header_css.headerContainer, style: headerContainerStyle, children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Tooltip.Tooltip, { text: headerContent, disabled: !isTruncated, children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
        "p",
        {
          className: headerClassName,
          style: (0, import_dynamic.assignInlineVars)({ [import_Header_css.maxLinesVar]: String(showedLines) }),
          children: [
            shownText,
            durationContent && durationContent !== "0" && /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { className: import_Header_css.headerSubText, children: durationSuffix })
          ]
        }
      ) }),
      shownHeader2Text && isHeader2Unique && /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Tooltip.Tooltip, { text: header2Content, disabled: !isHeader2Truncated, children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
        "p",
        {
          className: headerClassName,
          style: (0, import_dynamic.assignInlineVars)({
            [import_Header_css.maxLinesVar]: String(showedLines)
          }),
          children: [
            shownHeader2Text,
            duration2Content && duration2Content !== "0" && /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { className: import_Header_css.headerSubText, children: duration2Suffix })
          ]
        }
      ) }) }),
      shownSubheaderText && /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Tooltip.Tooltip, { text: subheaderContent, disabled: !isSubheaderTruncated, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("p", { className: import_Header_css.headerSubText, children: shownSubheaderText }) }) })
    ] }),
    displayHeaderDivider && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Divider.Divider, { orientation: "horizontal", className: import_tooltip_css.divider }, "divider")
  ] });
};
