// @dynatrace/strato-components/filters/filter-field/suggestions/AriaLiveInfo.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var AriaLiveInfo_exports = {};
__export(AriaLiveInfo_exports, {
  AriaLiveInfo: () => AriaLiveInfo
});
module.exports = __toCommonJS(AriaLiveInfo_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_VisuallyHidden = require("../../../core/components/visually-hidden/VisuallyHidden.js");
var import_can_receive_default_text_format = require("../../../core/utils/can-receive-default-text-format.js");
var import_useFilterFieldContext = require("../hooks/useFilterFieldContext.js");
var import_messages = require("../messages.js");
const MAX_ANNOUNCED_SUGGESTIONS = 5;
const ARIA_LIVE_DEBOUNCE_MS = 600;
const DefaultSuggestionsAriaLiveInfo = ({
  suggestions
}) => {
  if (!suggestions.length) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("ul", { children: suggestions.slice(0, MAX_ANNOUNCED_SUGGESTIONS).map((suggestion, index) => {
    let suggestionText;
    if ((0, import_can_receive_default_text_format.canReceiveDefaultTextFormat)(suggestion.displayValue)) {
      suggestionText = suggestion.displayValue;
    } else if ((0, import_can_receive_default_text_format.canReceiveDefaultTextFormat)(suggestion.value)) {
      suggestionText = suggestion.value;
    }
    if (suggestionText) {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("li", { children: suggestionText }, `${suggestionText}-${index}`);
    }
    return null;
  }) });
};
const LogicalOperatorSuggestionAriaLiveInfo = ({
  hasLogicalOperators
}) => {
  if (!hasLogicalOperators) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("li", { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_react_intl.FormattedMessage,
    {
      ...import_messages.filterFieldMessages.suggestions.ariaLiveInfo.logicalOperator
    }
  ) });
};
const GeneralSuggestionsAriaLiveInfo = ({
  values
}) => values.map((value) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)("li", { children: value.displayValue }, value.value));
const RecentAndPinnedSuggestionsAriaLiveInfo = ({
  message,
  count
}) => {
  if (count === 0) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("li", { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_react_intl.FormattedMessage,
    {
      ...message,
      values: {
        value: count
      }
    }
  ) });
};
const extractSuggestions = (flattenedItems, limit = MAX_ANNOUNCED_SUGGESTIONS) => {
  const result = [];
  for (const item of flattenedItems) {
    if (result.length >= limit) {
      break;
    }
    if (item.kind === "suggestion") {
      result.push(item.data);
    }
  }
  return result;
};
const AriaLiveInfo = ({
  flattenedSuggestionItems
}) => {
  const {
    isValidSearch,
    pastedContentSuggestionValues,
    logicalOperatorSuggestionValues,
    recentSuggestionValues,
    pinnedSuggestionValues,
    searchSuggestionValues,
    parserConfig
  } = (0, import_useFilterFieldContext.useFilterFieldContext)();
  const showSearchSuggestions = isValidSearch?.current && searchSuggestionValues.length > 0 && // eslint-disable-next-line @typescript-eslint/no-deprecated
  (parserConfig?.searchConversion ?? false);
  const hasAdditionalContent = pastedContentSuggestionValues.length > 0 || logicalOperatorSuggestionValues.length > 0 || showSearchSuggestions || recentSuggestionValues.length > 0 || pinnedSuggestionValues.length > 0;
  const combinedSuggestions = (0, import_react.useMemo)(
    () => extractSuggestions(flattenedSuggestionItems, MAX_ANNOUNCED_SUGGESTIONS),
    [flattenedSuggestionItems]
  );
  const [debouncedState, setDebouncedState] = (0, import_react.useState)(
    () => ({
      combinedSuggestions: [],
      hasAdditionalContent: false,
      showSearchSuggestions: false,
      pastedContentSuggestionValues: [],
      logicalOperatorSuggestionsAvailable: false,
      recentSuggestionCount: 0,
      pinnedSuggestionCount: 0,
      searchSuggestionValues: []
    })
  );
  (0, import_react.useEffect)(() => {
    const handle = setTimeout(() => {
      setDebouncedState({
        combinedSuggestions,
        hasAdditionalContent,
        showSearchSuggestions: !!showSearchSuggestions,
        pastedContentSuggestionValues,
        logicalOperatorSuggestionsAvailable: logicalOperatorSuggestionValues.length > 0,
        recentSuggestionCount: recentSuggestionValues.length,
        pinnedSuggestionCount: pinnedSuggestionValues.length,
        searchSuggestionValues
      });
    }, ARIA_LIVE_DEBOUNCE_MS);
    return () => {
      clearTimeout(handle);
    };
  }, [
    combinedSuggestions,
    hasAdditionalContent,
    showSearchSuggestions,
    pastedContentSuggestionValues,
    logicalOperatorSuggestionValues.length,
    recentSuggestionValues.length,
    pinnedSuggestionValues.length,
    searchSuggestionValues
  ]);
  const hasContent = debouncedState.combinedSuggestions.length > 0;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_VisuallyHidden.VisuallyHidden, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { "aria-live": "polite", "aria-atomic": "true", children: hasContent || debouncedState.hasAdditionalContent ? /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_react_intl.FormattedMessage,
      {
        ...import_messages.filterFieldMessages.suggestions.ariaLiveInfo.availableSuggestions
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      DefaultSuggestionsAriaLiveInfo,
      {
        suggestions: debouncedState.combinedSuggestions
      }
    ),
    debouncedState.hasAdditionalContent && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("ul", { children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        GeneralSuggestionsAriaLiveInfo,
        {
          values: debouncedState.pastedContentSuggestionValues
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        LogicalOperatorSuggestionAriaLiveInfo,
        {
          hasLogicalOperators: debouncedState.logicalOperatorSuggestionsAvailable
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        RecentAndPinnedSuggestionsAriaLiveInfo,
        {
          message: import_messages.filterFieldMessages.suggestions.ariaLiveInfo.recentlyUsedSuggestions,
          count: debouncedState.recentSuggestionCount
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        RecentAndPinnedSuggestionsAriaLiveInfo,
        {
          message: import_messages.filterFieldMessages.suggestions.ariaLiveInfo.pinnedSuggestions,
          count: debouncedState.pinnedSuggestionCount
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        GeneralSuggestionsAriaLiveInfo,
        {
          values: debouncedState.searchSuggestionValues
        }
      )
    ] }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_react_intl.FormattedMessage,
      {
        ...import_messages.filterFieldMessages.suggestions.ariaLiveInfo.moreSuggestionsAvailable
      }
    )
  ] }) : null }) });
};
