// @dynatrace/strato-components/filters/filter-field/suggestions/FallbackKeySuggestions.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var FallbackKeySuggestions_exports = {};
__export(FallbackKeySuggestions_exports, {
  FallbackKeySuggestionGroup: () => FallbackKeySuggestionGroup
});
module.exports = __toCommonJS(FallbackKeySuggestions_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_CSSHighlight = require("../components/CSSHighlight.js");
var import_useFilterFieldContext = require("../hooks/useFilterFieldContext.js");
var import_messages = require("../messages.js");
var suggestionStyles = __toESM(require("./Suggestion.sty.js"));
var import_Suggestion = require("./Suggestion.js");
var suggestionGroupStyles = __toESM(require("./SuggestionGroup.sty.js"));
var import_SuggestionGroup = require("./SuggestionGroup.js");
function FallbackKeySuggestionGroup() {
  const { fallbackKeySuggestionValues, statementSuggestionValues, term } = (0, import_useFilterFieldContext.useFilterFieldContext)();
  const intl = (0, import_react_intl.useIntl)();
  if (fallbackKeySuggestionValues.length === 0 && statementSuggestionValues.length === 0) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_SuggestionGroup.SuggestionGroup,
    {
      "aria-label": intl.formatMessage(
        import_messages.filterFieldMessages.suggestions.a11y.statementSuggestions
      ),
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_CSSHighlight.CSSHighlight,
          {
            className: suggestionGroupStyles.builtInSuggestionsContainer,
            term: term ?? "",
            selector: `.${suggestionStyles.statementSuggestionValue}`,
            highlightName: "filterfield-suggestions-highlight",
            children: statementSuggestionValues?.map((suggestion, index) => /* @__PURE__ */ (0, import_react.createElement)(
              import_Suggestion.Suggestion,
              {
                ...suggestion,
                key: suggestion.value ?? `${index}-statementsuggestions`,
                "data-strato-filter-field-virtual-focus-type": "statement-suggestion"
              }
            ))
          }
        ),
        fallbackKeySuggestionValues?.map((suggestion, index) => /* @__PURE__ */ (0, import_react.createElement)(
          import_Suggestion.Suggestion,
          {
            ...suggestion,
            key: suggestion.value ?? `${index}-fallbackkeysuggestions`,
            className: suggestionStyles.defaultSuggestion,
            "data-strato-filter-field-virtual-focus-type": "fallbackkey-suggestion"
          }
        ))
      ]
    }
  );
}
