// @dynatrace/strato-components/filters/filter-field/suggestions/RecentPinnedSuggestionItem.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var RecentPinnedSuggestionItem_exports = {};
__export(RecentPinnedSuggestionItem_exports, {
  PinnedSuggestionGroup: () => PinnedSuggestionGroup,
  RecentSuggestionGroup: () => RecentSuggestionGroup
});
module.exports = __toCommonJS(RecentPinnedSuggestionItem_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_Button = require("../../../buttons/button/Button.js");
var import_useId = require("../../../core/hooks/useId.js");
var import_CSSHighlight = require("../components/CSSHighlight.js");
var import_FilterStatement = require("../components/FilterStatement.js");
var import_useFilterFieldContext = require("../hooks/useFilterFieldContext.js");
var import_useFilterFieldLoggerContext = require("../hooks/useFilterFieldLoggerContext.js");
var import_useShowMore = require("../hooks/useShowMore.js");
var import_messages = require("../messages.js");
var styles = __toESM(require("./RecentPinnedSuggestionItem.sty.js"));
var suggestionStyles = __toESM(require("./Suggestion.sty.js"));
var import_Suggestion = require("./Suggestion.js");
var import_SuggestionGroup = require("./SuggestionGroup.js");
var import_SuggestionGroupLabel = require("./SuggestionGroupLabel.js");
var import_SuggestionPin = require("./SuggestionPin.js");
function RecentAndPinnedSuggestionItem(props) {
  const { suggestion, index, ranges, ...remainingProps } = props;
  const suggestionRef = (0, import_react.useRef)(null);
  const id = (0, import_useId.useId)();
  const { logger } = (0, import_useFilterFieldLoggerContext.useFilterFieldLoggerContext)();
  const suggestionHighlight = (0, import_react.useMemo)(
    () => (0, import_CSSHighlight.getOrCreateNamedHighlight)("filterfield-suggestions-highlight"),
    []
  );
  (0, import_react.useEffect)(() => {
    const highlightRanges = [];
    if (suggestionRef.current) {
      const valueElement = suggestionRef.current.querySelector(
        `.${suggestionStyles.value} span`
      );
      if (valueElement && valueElement.textContent !== "") {
        const textNode = valueElement.childNodes.item(0);
        if (textNode.nodeType === Node.TEXT_NODE) {
          const textLength = textNode.textContent?.length ?? 0;
          for (const [start, end] of ranges) {
            const clampedStart = Math.min(start, textLength);
            const clampedEnd = Math.min(end, textLength);
            if (clampedStart >= clampedEnd) {
              continue;
            }
            try {
              const range = new Range();
              range.setStart(textNode, clampedStart);
              range.setEnd(textNode, clampedEnd);
              suggestionHighlight.add(range);
              highlightRanges.push(range);
            } catch (error) {
              logger?.warn(
                "FilterField suggestion highlight range error",
                error
              );
            }
          }
        }
      }
    }
    return () => {
      if (highlightRanges) {
        for (const range of highlightRanges) {
          suggestionHighlight.delete(range);
        }
      }
    };
  }, [logger, ranges, suggestionHighlight]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_Suggestion.Suggestion,
    {
      className: styles.recentSuggestion,
      ref: suggestionRef,
      ...suggestion,
      ...remainingProps,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_FilterStatement.FilterStatement,
          {
            value: suggestion.displayValue?.toString(),
            highlighting: "minimal"
          }
        ),
        suggestion.value ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_SuggestionPin.SuggestionPin,
          {
            role: "option",
            filterString: suggestion.value,
            className: styles.suggestionPin,
            placement: "right",
            fallbackPlacements: ["right", "left"],
            id,
            "data-strato-filter-field-virtual-focus-id": id,
            "data-strato-filter-field-virtual-focus-type": "pinned-suggestion"
          }
        ) : null
      ]
    },
    `${suggestion.value ?? "non-interactive-"}${index}`
  );
}
function RecentSuggestionGroup() {
  const { filterFieldId, recentSuggestionValues } = (0, import_useFilterFieldContext.useFilterFieldContext)();
  let primaryCount = 2;
  if (recentSuggestionValues.length === 3) {
    primaryCount = 3;
  }
  const {
    items: slicedRecentSuggestionValues,
    isSliced,
    controlProps
  } = (0, import_useShowMore.useShowMore)({
    items: recentSuggestionValues,
    primaryCount
  });
  if (recentSuggestionValues.length === 0) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      import_SuggestionGroup.SuggestionGroup,
      {
        "aria-labelledby": `${filterFieldId ?? ""}-recent-filters`,
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_SuggestionGroupLabel.SuggestionGroupLabel, { id: `${filterFieldId ?? ""}-recent-filters`, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_intl.FormattedMessage, { ...import_messages.filterFieldMessages.suggestions.recentlyUsed }) }),
          slicedRecentSuggestionValues?.map((recentItem, index) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            RecentAndPinnedSuggestionItem,
            {
              suggestion: recentItem.suggestion,
              index,
              "data-strato-filter-field-virtual-focus-type": "recent-suggestion",
              ranges: recentItem.ranges
            },
            recentItem.suggestion.value ?? index
          ))
        ]
      }
    ),
    isSliced && /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { role: "presentation", children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button, { ...controlProps }) })
  ] });
}
function PinnedSuggestionGroup() {
  const { filterFieldId, pinnedSuggestionValues } = (0, import_useFilterFieldContext.useFilterFieldContext)();
  if (pinnedSuggestionValues.length === 0) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_SuggestionGroup.SuggestionGroup, { "aria-labelledby": `${filterFieldId ?? ""}-pinned-filters`, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_SuggestionGroupLabel.SuggestionGroupLabel, { id: `${filterFieldId ?? ""}-pinned-filters`, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_intl.FormattedMessage, { ...import_messages.filterFieldMessages.suggestions.pinnedFilters }) }),
    pinnedSuggestionValues?.map((pinnedItem, index) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      RecentAndPinnedSuggestionItem,
      {
        suggestion: pinnedItem.suggestion,
        index,
        "data-strato-filter-field-virtual-focus-type": "pinned-suggestion",
        ranges: pinnedItem.ranges
      },
      pinnedItem.suggestion.value ?? index
    ))
  ] });
}
