// @dynatrace/strato-components/filters/filter-field/suggestions/Suggestions.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Suggestions_exports = {};
__export(Suggestions_exports, {
  Suggestions: () => Suggestions
});
module.exports = __toCommonJS(Suggestions_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_virtual = require("@tanstack/react-virtual");
var import_clsx = __toESM(require("clsx"));
var import_framer_motion = require("framer-motion");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_react_popper = require("react-popper");
var import_useOverlayTransition = __toESM(require("../hooks/useOverlayTransition.js"));
var import_useShowMore = require("../hooks/useShowMore.js");
var import_messages = require("../messages.js");
var import_AriaLiveInfo = require("./AriaLiveInfo.js");
var import_FallbackKeySuggestions = require("./FallbackKeySuggestions.js");
var import_LogicalOperatorSuggestions = require("./LogicalOperatorSuggestions.js");
var import_PastedContentSuggestions = require("./PastedContentSuggestions.js");
var import_RecentPinnedSuggestionItem = require("./RecentPinnedSuggestionItem.js");
var import_SearchSuggestions = require("./SearchSuggestions.js");
var suggestionStyles = __toESM(require("./Suggestion.sty.js"));
var import_Suggestion = require("./Suggestion.js");
var suggestionGroupStyles = __toESM(require("./SuggestionGroup.sty.js"));
var import_SuggestionGroupLabel = require("./SuggestionGroupLabel.js");
var styles = __toESM(require("./SuggestionsOverlay.sty.js"));
var import_Button = require("../../../buttons/button/Button.js");
var import_ProgressCircle = require("../../../content/progress/ProgressCircle.js");
var import_VisuallyHidden = require("../../../core/components/visually-hidden/VisuallyHidden.js");
var import_useBehavioralTrackingProps = require("../../../core/hooks/useBehavioralTrackingProps.js");
var import_Flex = require("../../../layouts/flex/Flex.js");
var import_Surface = require("../../../layouts/surface/Surface.js");
var import_CSSHighlight = require("../components/CSSHighlight.js");
var import_useFilterFieldContext = require("../hooks/useFilterFieldContext.js");
var import_flatten_suggestions = require("../utils/suggestions-utils/flatten-suggestions.js");
function toGroupLabelId(groupId) {
  return `group-label-${groupId.replace(/\s+/g, "-")}`;
}
function renderFlattenedSuggestionItem(flatItem, trailing) {
  switch (flatItem.kind) {
    case "group-header": {
      let child;
      if ((0, import_react.isValidElement)(flatItem.label)) {
        child = flatItem.label;
      } else if (flatItem.label) {
        child = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_SuggestionGroupLabel.SuggestionGroupLabel, { children: flatItem.label });
      } else {
        child = null;
      }
      return {
        child: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          "div",
          {
            className: suggestionGroupStyles.groupHeaderSeparator,
            role: "presentation",
            children: child
          }
        ),
        key: `group-header-${flatItem.groupId}`
      };
    }
    case "suggestion": {
      const ariaDescribedBy = flatItem.groupId ? toGroupLabelId(flatItem.groupId) : void 0;
      return {
        child: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_Suggestion.Suggestion,
          {
            ...flatItem.data,
            trailing: flatItem.data.trailing ?? trailing,
            "aria-describedby": ariaDescribedBy
          }
        ),
        key: `${flatItem.data.value ?? flatItem.data.displayValue ?? ""}`
      };
    }
    case "react-node": {
      const groupLabelId = flatItem.groupId ? toGroupLabelId(flatItem.groupId) : void 0;
      let child = flatItem.node;
      let key = "";
      if ((0, import_react.isValidElement)(flatItem.node) && groupLabelId) {
        const existing = flatItem.node.props["aria-describedby"] ?? "";
        const merged = existing ? `${existing} ${groupLabelId}` : groupLabelId;
        child = (0, import_react.cloneElement)(flatItem.node, {
          "aria-describedby": merged
        });
        key = `react-node-${flatItem.node.key}`;
      }
      return { child, key };
    }
  }
}
function SuggestionContent(props) {
  const { optionVirtualizer, flattenedSuggestions, isSliced, controlProps } = props;
  const intl = (0, import_react_intl.useIntl)();
  const {
    term,
    editorView,
    autoSuggestions,
    recentSuggestionValues,
    pinnedSuggestionValues,
    logicalOperatorSuggestionValues,
    pastedContentSuggestionValues,
    searchSuggestionValues,
    fallbackKeySuggestionValues,
    statementSuggestionValues,
    filterFieldId,
    parserConfig,
    isValidSearch
  } = (0, import_useFilterFieldContext.useFilterFieldContext)();
  const { trailing } = autoSuggestions;
  const showLogicalOperators = logicalOperatorSuggestionValues.length > 0;
  const showRecents = recentSuggestionValues.length > 0;
  const showPinned = pinnedSuggestionValues.length > 0;
  const showFallbackKeySuggestions = fallbackKeySuggestionValues.length > 0;
  const showStatementSuggestions = statementSuggestionValues.length > 0;
  const showSearchSuggestions = isValidSearch?.current && searchSuggestionValues.length > 0 && // eslint-disable-next-line @typescript-eslint/no-deprecated
  (parserConfig?.searchConversion ?? false);
  const showPastedContentSuggestions = pastedContentSuggestionValues.length > 0;
  const showBuiltInSuggestions = showRecents || showPinned || showLogicalOperators || showPastedContentSuggestions || showSearchSuggestions || showFallbackKeySuggestions || showStatementSuggestions;
  const isEmpty = editorView?.state.doc.toString() === "";
  const childrenCount = flattenedSuggestions.length;
  if (childrenCount === 0 && !showBuiltInSuggestions) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Suggestion.Suggestion,
      {
        trailing: "",
        displayValue: intl.formatMessage(
          import_messages.filterFieldMessages.suggestions.noSuggestions
        )
      }
    );
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { id: `suggestions-${filterFieldId}`, children: [
    childrenCount > 0 ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_CSSHighlight.CSSHighlight,
      {
        term: term ?? "",
        selector: `.${suggestionStyles.value}:not(:has(mark)), .${suggestionStyles.matchableSuggestionDetails}:not(:has(mark))`,
        highlightName: "filterfield-suggestions-highlight",
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          "div",
          {
            style: {
              width: "100%",
              position: "relative",
              height: optionVirtualizer.getTotalSize(),
              display: "grid",
              gridTemplateRows: optionVirtualizer.measurementsCache.map((el) => `${el.size}px`).join(" ")
            },
            children: optionVirtualizer.getVirtualItems().map((virtualRow) => {
              const flatItem = flattenedSuggestions[virtualRow.index];
              const { child, key: suggestionKey } = renderFlattenedSuggestionItem(flatItem, trailing);
              return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                "div",
                {
                  className: (0, import_clsx.default)(
                    "strato-filter-field-suggestion",
                    suggestionStyles.virtualizedSuggestion
                  ),
                  "data-index": virtualRow.index,
                  style: {
                    gridRow: virtualRow.index + 1,
                    minWidth: 0,
                    height: "fit-content"
                  },
                  ref: (el) => {
                    optionVirtualizer.measureElement(el);
                  },
                  children: child
                },
                `${virtualRow.key}-${suggestionKey}`
              );
            })
          },
          `wrapper-${childrenCount}`
        )
      }
    ) : null,
    isSliced && /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { role: "presentation", children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button, { ...controlProps }) }),
    showBuiltInSuggestions && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_PastedContentSuggestions.PastedContentSuggestionGroup, {}),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_FallbackKeySuggestions.FallbackKeySuggestionGroup, {}),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_LogicalOperatorSuggestions.LogicalOperatorSuggestionsGroup, {}),
      showSearchSuggestions && !isEmpty && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_SearchSuggestions.SearchSuggestionGroup, {}),
      /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
        import_CSSHighlight.CSSHighlight,
        {
          className: suggestionGroupStyles.builtInSuggestionsContainer,
          term: editorView?.state.doc.toString() ?? "",
          selector: `.${suggestionStyles.value}`,
          highlightName: "filterfield-suggestions-highlight",
          children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_RecentPinnedSuggestionItem.RecentSuggestionGroup, {}),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_RecentPinnedSuggestionItem.PinnedSuggestionGroup, {})
          ]
        }
      ),
      showSearchSuggestions && isEmpty && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_SearchSuggestions.SearchSuggestionGroup, {})
    ] })
  ] });
}
const heightModifier = {
  name: "filterfield-height-handler",
  enabled: true,
  fn: ({ state }) => {
    const viewportHeight = Math.max(
      document.documentElement.clientHeight || 0,
      window.innerHeight || 0
    );
    const maxHeightBase = Math.max(viewportHeight / 2, 320);
    state.styles.popper.maxHeight = `${maxHeightBase - state.rects.reference.height - 20}px`;
  },
  phase: "beforeWrite",
  requires: ["computeStyles"]
};
const suggestionsAnimationVariants = {
  inital: {
    translateY: "-10px",
    // Compensate the parent scaling the content to 0.8. Due to round errors when the height gets taller
    // the content sometimes got larger than the intended which lead to a "vertigo" effect. By adding
    // 0.05 we can counteract that effect somewhat.
    scaleY: 1 / 0.85,
    opacity: 0
  },
  default: {
    translateY: 0,
    scaleY: 1,
    opacity: 1
  },
  exit: {
    translateY: "-5px",
    opacity: 0
  }
};
function getSuggestionsCount(defaultSuggestionsCount, isEmpty) {
  if (typeof defaultSuggestionsCount === "number") {
    return defaultSuggestionsCount;
  }
  return defaultSuggestionsCount[isEmpty ? "empty" : "filled"];
}
const Suggestions = (props) => {
  const {
    loading,
    children,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const { autoSuggestions, defaultSuggestionsCount, triggerRef, editorView } = (0, import_useFilterFieldContext.useFilterFieldContext)();
  const intl = (0, import_react_intl.useIntl)();
  const transition = (0, import_useOverlayTransition.default)(true);
  const parentRef = (0, import_react.useRef)(null);
  const childrenArray = (0, import_react.useMemo)(() => import_react.Children.toArray(children), [children]);
  const { flattenedSuggestionItems, groupMetadata } = (0, import_react.useMemo)(
    () => (0, import_flatten_suggestions.flattenSuggestions)([
      ...Object.values(autoSuggestions.groupedSuggestions ?? {}).filter(
        (group) => group && (group.suggestions?.length ?? 0) > 0
      ),
      ...childrenArray
    ]),
    [autoSuggestions.groupedSuggestions, childrenArray]
  );
  const isEmpty = editorView?.state.doc.toString() === "";
  const suggestionsCount = getSuggestionsCount(
    defaultSuggestionsCount,
    isEmpty
  );
  const {
    items: slicedSuggestions,
    isSliced,
    controlProps
  } = (0, import_useShowMore.useShowMore)({
    items: flattenedSuggestionItems,
    primaryCount: suggestionsCount,
    sliceFn: import_useShowMore.sliceFlattenedSuggestions
  });
  const optionVirtualizer = (0, import_react_virtual.useVirtualizer)({
    count: slicedSuggestions.length,
    getScrollElement: () => parentRef.current,
    overscan: 5,
    estimateSize: () => 32
  });
  const suggestionsRef = (0, import_react.useRef)(null);
  const { styles: positioningStyles } = (0, import_react_popper.usePopper)(
    triggerRef,
    suggestionsRef.current,
    {
      placement: "bottom-start",
      modifiers: [heightModifier]
    }
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_Surface.Surface,
    {
      className: (0, import_clsx.default)(consumerClassName, styles.suggestionsOverlay),
      style: consumerStyle,
      elevation: "floating",
      ref: suggestionsRef,
      "data-testid": dataTestId,
      ...behavioralTrackingProps,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_AriaLiveInfo.AriaLiveInfo, { flattenedSuggestionItems }),
        groupMetadata.length > 0 && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_VisuallyHidden.VisuallyHidden, { children: groupMetadata.map(({ groupId, label }) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { id: toGroupLabelId(groupId), children: label }, groupId)) }),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          "div",
          {
            className: styles.suggestionsScrollContainerWrapper,
            role: "listbox",
            "aria-expanded": "true",
            "aria-label": intl.formatMessage(
              import_messages.filterFieldMessages.suggestions.a11y.ariaLabel
            ),
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_framer_motion.motion.div,
              {
                className: (0, import_clsx.default)(
                  styles.suggestionsScrollContainer,
                  "strato-filter-field-suggestions"
                ),
                ref: parentRef,
                style: { maxHeight: positioningStyles.popper.maxHeight },
                variants: suggestionsAnimationVariants,
                initial: "inital",
                animate: "default",
                exit: "exit",
                transition,
                children: loading ? /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
                  import_Flex.Flex,
                  {
                    role: "presentation",
                    justifyContent: "start",
                    px: 8,
                    py: 6,
                    className: suggestionStyles.suggestionsLoading,
                    children: [
                      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ProgressCircle.ProgressCircle, { size: "small", "aria-label": "Loading suggestions" }),
                      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_intl.FormattedMessage, { ...import_messages.filterFieldMessages.suggestions.loading })
                    ]
                  }
                ) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  SuggestionContent,
                  {
                    optionVirtualizer,
                    flattenedSuggestions: slicedSuggestions,
                    isSliced,
                    controlProps
                  }
                )
              }
            )
          }
        )
      ]
    }
  );
};
Suggestions.displayValue = "FilterField.Suggestions";
