// @dynatrace/strato-components/filters/filter-field/suggestions/suggestions-extension.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var suggestions_extension_exports = {};
__export(suggestions_extension_exports, {
  getTokenBeforePastedContent: () => getTokenBeforePastedContent,
  suggestionsExtension: () => suggestionsExtension
});
module.exports = __toCommonJS(suggestions_extension_exports);
var import_autocomplete = require("@codemirror/autocomplete");
var import_language = require("@codemirror/language");
var import_view = require("@codemirror/view");
var import_editor_config = require("../editor-state-configs/editor-config.js");
var import_escape_suggestion = require("../utils/escape-suggestion.js");
var import_get_current_token = require("../utils/suggestions-utils/get-current-token.js");
var import_get_suggestion_category = require("../utils/suggestions-utils/get-suggestion-category.js");
var import_get_suggestions = require("../utils/suggestions-utils/get-suggestions.js");
var import_key_suggestions = require("../utils/suggestions-utils/key-suggestions.js");
var import_recent_and_pinned_suggestions = require("../utils/suggestions-utils/recent-and-pinned-suggestions.js");
var import_suggestions_callback = require("../utils/suggestions-utils/suggestions-callback.js");
var import_tree_navigation = require("../utils/tree-navigation.js");
const suggestionsExtension = ({
  overlayState,
  numberParser,
  internalValidatorMap,
  onSuggest,
  _onSuggest,
  getSyntaxTree,
  customTypes,
  parserConfig
}) => {
  const filterFieldSuggestion = (context) => {
    if (context.view?.composing) {
      return null;
    }
    const tree = (0, import_language.syntaxTree)(context.state);
    const { tree: convertedTree } = getSyntaxTree({
      tree,
      numberParser,
      doc: context.state.doc.toString(),
      customTypes,
      parserConfig
    });
    const pastingStateField = context.view?.state.field(import_editor_config.pastingMode, false);
    const { head: cursorPosition } = context.state.selection.main;
    let currentStatement = (0, import_suggestions_callback.getFilterStatement)(
      convertedTree,
      cursorPosition,
      context.state.doc.toString()
    );
    const token = (0, import_get_current_token.getCurrentToken)({
      tree,
      convertedTree,
      cursorPosition,
      doc: context.state.doc.toString()
    });
    let currentToken = token?.currentToken;
    const currentTokenType = token?.currentTokenType;
    const treeCursor = tree.cursorAt(cursorPosition, -1);
    treeCursor.childBefore(cursorPosition);
    const prevNode = treeCursor.node.type.is("FilterValue") || treeCursor.node.parent?.type.is("FilterValue") ? "FilterValue" : treeCursor.node.name;
    const doc = context.state.doc;
    const {
      showInclusionOperators,
      showLogicalOperators,
      suggestionType,
      overlappingSuggestions
    } = (0, import_get_suggestion_category.getSuggestionCategory)({
      currentStatement,
      currentToken,
      currentTokenType,
      prevNode,
      tree,
      doc: context.state.doc.toString(),
      cursorPosition
    });
    const suggestionTypes = Array.from(
      new Set(
        [suggestionType].concat(
          // Using concat with the ternary so that typescript can correctly infer the array's type.
          overlappingSuggestions ? overlappingSuggestions.type : []
        )
      )
    );
    let suggestionTypeForToken = suggestionTypes.map((type) => ({
      type,
      token: type === overlappingSuggestions?.type ? overlappingSuggestions.subToken : currentToken
    }));
    if (pastingStateField && pastingStateField.pastingInProgress && pastingStateField.range) {
      suggestionTypes.push("pastedContent");
      currentStatement = void 0;
      const textValue = doc.sliceString(
        pastingStateField.range.from,
        pastingStateField.range.to
      );
      const escapedTextValue = (0, import_escape_suggestion.escapeSuggestion)(textValue);
      const pastedContentToken = {
        range: {
          from: pastingStateField.range.from,
          to: pastingStateField.range.to
        },
        value: textValue,
        textValue: escapedTextValue,
        type: "String"
      };
      const { token: token2, type } = getCurrentTokenForPastedContent(
        currentToken,
        pastingStateField.range,
        context.state,
        pastedContentToken
      );
      if (type) {
        suggestionTypeForToken = [
          {
            type,
            token: token2
          }
        ];
        currentToken = token2;
      }
      suggestionTypeForToken.push({
        type: "pastedContent",
        token: pastedContentToken
      });
    }
    const {
      logicalOperatorSuggestions,
      groupedSuggestions,
      pastedContentSuggestions
    } = (0, import_get_suggestions.getSuggestions)({
      suggestionTypes,
      overlappingSuggestions,
      currentStatement,
      currentToken,
      currentTokenType,
      prevNode,
      internalValidatorMap,
      customTypes,
      showLogicalOperators,
      showInclusionOperators,
      doc: context.state.doc.toString(),
      suggestionTypeForToken
    });
    const { recentSuggestionValues, pinnedSuggestionValues } = (0, import_recent_and_pinned_suggestions.getRecentAndPinnedSuggestions)(context);
    const searchSuggestionValues = (0, import_key_suggestions.getSearchSuggestions)(
      context,
      currentStatement,
      currentToken,
      suggestionTypes,
      (0, import_tree_navigation.isCursorInsideGroup)(context.state, cursorPosition)
    );
    let filterStatementSuggestions = [];
    if (currentToken.type !== "List") {
      filterStatementSuggestions = (0, import_key_suggestions.getFilterStatementSuggestions)(
        currentToken,
        suggestionTypes,
        internalValidatorMap?.filterStatementKeysMap
      );
    }
    const fallbackKeySuggestions = (0, import_key_suggestions.getFallbackKeySuggestions)(
      currentToken,
      suggestionTypes,
      internalValidatorMap?.fallbackKeys,
      filterStatementSuggestions
    );
    _onSuggest({
      suggestionsContext: {
        token: currentToken,
        currentStatement
      },
      autoSuggestionValues: groupedSuggestions,
      recentSuggestionValues,
      pinnedSuggestionValues,
      logicalOperatorSuggestionValues: logicalOperatorSuggestions,
      pastedContentSuggestionValues: pastedContentSuggestions,
      searchSuggestionValues,
      fallbackKeySuggestionValues: fallbackKeySuggestions,
      filterStatementSuggestionValues: filterStatementSuggestions
    });
    onSuggest?.({
      suggestionTypes,
      currentTokens: suggestionTypeForToken,
      value: context.state.doc.toString(),
      groupedSuggestions,
      syntaxTree: convertedTree,
      cursorPosition,
      currentStatement
    });
    overlayState.open();
    return null;
  };
  return [
    (0, import_autocomplete.autocompletion)({ override: [filterFieldSuggestion] }),
    import_view.EditorView.updateListener.of((update) => {
      if (update.view.composing) {
        return;
      }
      if (update.focusChanged && update.view.hasFocus || update.transactions.some(
        (transaction) => transaction.isUserEvent("input") || transaction.isUserEvent(import_editor_config.ANNOTATION_SUGGESTION_APPLY) || transaction.isUserEvent(import_editor_config.ANNOTATION_SUGGESTION_PIN) || transaction.isUserEvent("delete") || transaction.selection !== void 0 && (transaction.isUserEvent(import_editor_config.ANNOTATION_DOC_CHANGED_CONTROLLED) || update.view.hasFocus)
      )) {
        (0, import_autocomplete.startCompletion)(update.view);
        return;
      }
      if (!update.view.hasFocus) {
        overlayState.close();
        return;
      }
    })
  ];
};
function getCurrentTokenForPastedContent(currentToken, pastingRange, state, pastedContentToken) {
  if (pastingRange.from >= currentToken.range.from && pastingRange.to <= currentToken.range.to) {
    return { token: currentToken, type: void 0 };
  }
  return getTokenBeforePastedContent(
    pastingRange.from,
    state,
    pastedContentToken
  );
}
function getTrailingNodeEnd(tree, position) {
  const trailingCursor = tree.cursorAt(position, 1);
  if ((trailingCursor.node.name === "FreeText" || trailingCursor.node.name === "Identifier") && trailingCursor.node.from <= position) {
    return trailingCursor.node.to;
  }
  return null;
}
function getTokenBeforePastedContent(pastedRangeFrom, editorState, pastedContentToken) {
  const tree = (0, import_language.syntaxTree)(editorState);
  const cursor = tree.cursorAt(pastedRangeFrom, -1);
  if (cursor.node.name === "Filter") {
    return { token: pastedContentToken, type: "key" };
  }
  if (cursor.node.name === "FilterStatement" || cursor.node.name === "ComparisonOperator") {
    const rangeTo = getTrailingNodeEnd(tree, pastedContentToken.range.to);
    if (rangeTo !== null) {
      const textValue = editorState.doc.sliceString(
        pastedContentToken.range.from,
        rangeTo
      );
      return {
        token: {
          range: { from: pastedContentToken.range.from, to: rangeTo },
          value: textValue,
          textValue: (0, import_escape_suggestion.escapeSuggestion)(textValue),
          type: "String"
        },
        type: "value"
      };
    }
    return { token: pastedContentToken, type: "value" };
  }
  if (cursor.node.name === "FreeText" || cursor.node.name === "Identifier") {
    const rangeFrom = cursor.node.from;
    const rangeTo = getTrailingNodeEnd(tree, pastedContentToken.range.to) ?? pastedContentToken.range.to;
    const textValue = editorState.doc.sliceString(rangeFrom, rangeTo);
    const combinedToken = {
      range: { from: rangeFrom, to: rangeTo },
      value: textValue,
      textValue: (0, import_escape_suggestion.escapeSuggestion)(textValue),
      type: "String"
    };
    return {
      token: combinedToken,
      type: cursor.node.name === "FreeText" ? "key" : "value"
    };
  }
  return { token: pastedContentToken, type: void 0 };
}
