// @dynatrace/strato-components/filters/filter-field/suggestions/Suggestion.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Suggestion_exports = {};
__export(Suggestion_exports, {
  Suggestion: () => Suggestion
});
module.exports = __toCommonJS(Suggestion_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_clsx = __toESM(require("clsx"));
var import_lodash_es = require("lodash");
var import_react2 = require("react");
var import_react_intl = require("react-intl");
var styles = __toESM(require("./Suggestion.sty.js"));
var import_SuggestionDetails = require("./SuggestionDetails.js");
var import_SuggestionPin = require("./SuggestionPin.js");
var import_Chip = require("../../../content/chip/Chip.js");
var import_useId = require("../../../core/hooks/useId.js");
var import_useMergeRefs = require("../../../core/hooks/useMergeRefs.js");
var import_useSlots = require("../../../core/hooks/useSlots.js");
var import_getVirtualFocusStyles = require("../../../core/styles/getVirtualFocusStyles.js");
var import_can_receive_default_text_format = require("../../../core/utils/can-receive-default-text-format.js");
var import_Flex = require("../../../layouts/flex/Flex.js");
var import_Grid = require("../../../layouts/grid/Grid.js");
var import_Tooltip = require("../../../overlays/tooltip/Tooltip.js");
var import_editor_config = require("../editor-state-configs/editor-config.js");
var import_useFilterFieldContext = require("../hooks/useFilterFieldContext.js");
var import_messages = require("../messages.js");
var import_filter_field_state_reducer = require("../state/filter-field-state-reducer.js");
function WrappedSuggestionLabel({
  displayLabel
}) {
  if (typeof displayLabel === "string" || (0, import_lodash_es.isArray)(displayLabel)) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { className: styles.suggestionText, children: displayLabel });
  }
  return displayLabel;
}
function getDescription(type) {
  switch (type) {
    case "Boolean":
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_react_intl.FormattedMessage,
        {
          ...import_messages.filterFieldMessages.suggestions.supportedValues.Boolean
        }
      );
    case "Duration":
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_react_intl.FormattedMessage,
        {
          ...import_messages.filterFieldMessages.suggestions.supportedValues.Durations
        }
      );
    case "Number":
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_react_intl.FormattedMessage,
        {
          ...import_messages.filterFieldMessages.suggestions.supportedValues.Numbers
        }
      );
    case "IPAddress":
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_react_intl.FormattedMessage,
        {
          ...import_messages.filterFieldMessages.suggestions.supportedValues.IPAddresses
        }
      );
    case "UID":
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_react_intl.FormattedMessage,
        {
          ...import_messages.filterFieldMessages.suggestions.supportedValues.UIDs
        }
      );
    case "SmartscapeId":
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_react_intl.FormattedMessage,
        {
          ...import_messages.filterFieldMessages.suggestions.supportedValues.SmartscapeIds
        }
      );
    case "Timestamp":
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_react_intl.FormattedMessage,
        {
          ...import_messages.filterFieldMessages.suggestions.supportedValues.Timestamps
        }
      );
    case "String":
    case "Any":
    default:
      return void 0;
  }
}
function dispatchEndPastingModeEffect(editorView) {
  editorView?.dispatch({
    effects: [
      import_editor_config.pastingEffect.of({
        pastingInProgress: false
      })
    ],
    // In order to re-run the decorations and re-generate errors and diagnostics,
    // we have to dispatch a new document.
    changes: {
      from: 0,
      to: editorView?.state.doc.length,
      insert: editorView?.state.doc
    },
    selection: editorView.state.selection,
    annotations: import_editor_config.stringAnnotation.of(import_editor_config.ANNOTATION_END_PASTING_MODE)
  });
}
function handleSuggestionApply(editorView, id, applySuggestion, suggestionRef, suggestionText, trailing, insertionStrategy, insertionRange, cursorBeforeTrailing, escapeSuggestion) {
  if (id === "strato-filterfield-suggestion-pasted-content") {
    dispatchEndPastingModeEffect(editorView);
  } else {
    const pastingInProgress = editorView?.state.field(import_editor_config.pastingMode, false);
    if (pastingInProgress?.pastingInProgress) {
      dispatchEndPastingModeEffect(editorView);
    }
    applySuggestion(
      suggestionRef,
      suggestionText,
      trailing,
      insertionStrategy,
      cursorBeforeTrailing,
      insertionRange,
      escapeSuggestion
    );
  }
}
const Suggestion = (0, import_react2.forwardRef)((props, forwardedRef) => {
  const {
    value,
    displayValue,
    details,
    trailing = " ",
    cursorBeforeTrailing = false,
    insertionStrategy = "replace-token",
    insertionRange,
    escapeSuggestion,
    valueType,
    className: consumerClassName,
    style: consumerStyle,
    icon,
    "data-testid": dataTestId,
    "aria-describedby": ariaDescribedBy,
    "aria-labelledby": ariaLabelledBy,
    children,
    id,
    suggestionType,
    ...remainingProps
  } = props;
  const suggestionRef = (0, import_react2.useRef)(null);
  const rootRef = (0, import_useMergeRefs.useMergeRefs)([forwardedRef, suggestionRef]);
  const { applySuggestion, dispatch, focusedListItem, editorView } = (0, import_useFilterFieldContext.useFilterFieldContext)();
  const valueRef = (0, import_react2.useRef)(null);
  const generatedId = (0, import_useId.useId)();
  const uniqueId = id ?? generatedId;
  const [slots, label] = (0, import_useSlots.useSlots)(children, {
    details: import_SuggestionDetails.SuggestionDetails,
    pin: import_SuggestionPin.SuggestionPin
  });
  const childIsText = (0, import_can_receive_default_text_format.canReceiveDefaultTextFormat)(label);
  const displayValueIsText = (0, import_can_receive_default_text_format.canReceiveDefaultTextFormat)(displayValue);
  const isFocused = uniqueId === focusedListItem?.id ? true : void 0;
  const finalDetails = (0, import_react2.useMemo)(() => {
    if (slots.details) {
      return (0, import_react2.cloneElement)(slots.details, {
        className: (0, import_clsx.default)(
          slots.details.props.className,
          styles.matchableSuggestionDetails
        )
      });
    }
    if (typeof details === "string") {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_SuggestionDetails.SuggestionDetails, { className: styles.matchableSuggestionDetails, children: details });
    }
    if (details) {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_SuggestionDetails.SuggestionDetails, { children: details });
    }
    if (suggestionType !== "validatorMap") {
      const description = getDescription(valueType);
      if (description) {
        return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_SuggestionDetails.SuggestionDetails, { children: description });
      }
    }
    return null;
  }, [details, slots.details, suggestionType, valueType]);
  const [showTooltip, setShowTooltip] = (0, import_react2.useState)(false);
  if (value === void 0 && !childIsText && !displayValueIsText) {
    console.warn(
      "Suggestion value: There is no value provided for a suggestion and the children are not text-only. The applied result may be unexpected."
    );
  }
  let suggestionText = "";
  if (value) {
    suggestionText = value;
  } else if (childIsText) {
    suggestionText = label.join("");
  } else if (displayValueIsText) {
    suggestionText = displayValue;
  }
  if (typeof suggestionText === "string" && !trailing) {
    suggestionText = suggestionText.trim();
  }
  (0, import_react2.useEffect)(() => {
    if (typeof suggestionText === "string") {
      dispatch({
        type: import_filter_field_state_reducer.FilterFieldActions.ADDSUGGESTION,
        payload: {
          id: uniqueId,
          applySuggestion: () => {
            handleSuggestionApply(
              editorView,
              uniqueId,
              applySuggestion,
              suggestionRef.current,
              suggestionText,
              trailing,
              insertionStrategy,
              insertionRange,
              cursorBeforeTrailing,
              escapeSuggestion
            );
          }
        }
      });
    }
    return () => dispatch({
      type: import_filter_field_state_reducer.FilterFieldActions.REMOVESUGGESTION,
      payload: { id: uniqueId }
    });
  }, []);
  if (typeof suggestionText !== "string") {
    return null;
  }
  const displayLabel = label.length > 0 ? label : displayValue ?? value;
  if (!value && trailing === "" && id !== "strato-filterfield-suggestion-pasted-content") {
    return /* @__PURE__ */ (0, import_react.createElement)(
      import_Flex.Flex,
      {
        ...remainingProps,
        role: "presentation",
        key: value,
        justifyContent: "space-between",
        className: id === "strato-filterfield-suggestion-jsonpath-hint" ? styles.suggestionsJSONPathHint : styles.suggestionsEmptyState,
        ref: rootRef
      },
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { children: displayLabel }),
      finalDetails
    );
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_Flex.Flex,
    {
      ...remainingProps,
      role: slots.pin && "group",
      gap: 0,
      "data-testid": dataTestId,
      className: (0, import_clsx.default)(styles.suggestion, consumerClassName),
      onTouchEnd: (e) => {
        e.preventDefault();
      },
      style: consumerStyle,
      ref: rootRef,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_Tooltip.Tooltip,
          {
            text: displayLabel,
            placement: "left",
            open: showTooltip,
            className: styles.tooltipWrapper,
            onOpenChange: (isOpen) => {
              const valueNodeChild = valueRef.current?.children[0];
              const hasOverFlow = (valueNodeChild?.scrollWidth ?? 0) > (valueNodeChild?.clientWidth ?? 0);
              setShowTooltip(isOpen && hasOverFlow);
            },
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
              import_Grid.Grid,
              {
                id: uniqueId,
                role: "option",
                "aria-selected": isFocused,
                "aria-labelledby": ariaLabelledBy,
                "aria-describedby": ariaDescribedBy,
                "data-strato-filter-field-virtual-focus-id": uniqueId,
                "data-strato-filter-field-virtual-focus-type": "suggestion",
                columnGap: 8,
                rowGap: 0,
                className: (0, import_clsx.default)(
                  styles.suggestionContent,
                  (0, import_getVirtualFocusStyles.getVirtualFocusStyles)(!!isFocused).className,
                  id === "strato-filterfield-suggestion-pasted-content" ? styles.pastingSuggestion : {}
                ),
                paddingLeft: icon ? 8 : 12,
                onPointerDown: () => {
                  handleSuggestionApply(
                    editorView,
                    id,
                    applySuggestion,
                    suggestionRef.current,
                    suggestionText,
                    trailing,
                    insertionStrategy,
                    insertionRange,
                    cursorBeforeTrailing,
                    escapeSuggestion
                  );
                },
                "data-dt-name": "Suggestion",
                children: [
                  /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { gap: 8, minWidth: 0, children: [
                    icon && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Chip.Chip, { size: "condensed", className: styles.icon, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Chip.Chip.Prefix, { children: icon }) }),
                    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                      import_Flex.Flex,
                      {
                        className: (0, import_clsx.default)(styles.value, {
                          [`${styles.truncateWithinMark}`]: id === "strato-filterfield-suggestion-use-as-value" || id === "strato-filterfield-suggestion-use-as-key" || id === "strato-filterfield-suggestion-pasted-content" || id === "strato-filterfield-suggestion-search"
                        }),
                        ref: valueRef,
                        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(WrappedSuggestionLabel, { displayLabel })
                      }
                    )
                  ] }),
                  finalDetails
                ]
              }
            )
          }
        ),
        slots.pin
      ]
    }
  );
});
Object.assign(Suggestion, { displayValue: "FilterField.Suggestion" });
