// @dynatrace/strato-components/filters/filter-field/suggestions/SuggestionTips.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SuggestionTips_exports = {};
__export(SuggestionTips_exports, {
  SuggestionTips: () => SuggestionTips
});
module.exports = __toCommonJS(SuggestionTips_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_framer_motion = require("framer-motion");
var import_react_intl = require("react-intl");
var styles = __toESM(require("./SuggestionTips.sty.js"));
var import_Flex = require("../../../layouts/flex/Flex.js");
var import_Surface = require("../../../layouts/surface/Surface.js");
var import_textStyle_css = require("../../../styles/textStyle.sty.js");
var import_ExternalLink = require("../../../typography/external-link/ExternalLink.js");
var import_Text = require("../../../typography/text/Text.js");
var import_TextEllipsis = require("../../../typography/text-ellipsis/TextEllipsis.js");
var import_useOverlayTransition = __toESM(require("../hooks/useOverlayTransition.js"));
var import_messages = require("../messages.js");
const suggestionTipsVariants = {
  enter: {
    // Compensate the parent scaling the content to 0.8
    scaleY: 1 / 0.85,
    opacity: 0
  },
  default: {
    scaleY: 1,
    opacity: 1
  },
  exit: {
    opacity: 0
  }
};
const SuggestionTips = () => {
  const intl = (0, import_react_intl.useIntl)();
  const transition = (0, import_useOverlayTransition.default)(true);
  const tips = [
    {
      shortcut: "\u23CE",
      text: intl.formatMessage(import_messages.filterFieldMessages.suggestions.tips.apply)
    },
    {
      shortcut: "\u2193\u2191",
      text: intl.formatMessage(import_messages.filterFieldMessages.suggestions.tips.selection)
    },
    {
      shortcut: "esc",
      text: intl.formatMessage(import_messages.filterFieldMessages.suggestions.tips.close)
    }
  ];
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Surface.Surface,
    {
      className: styles.suggestionTips,
      paddingX: 16,
      paddingY: 8,
      elevation: "floating",
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
        import_Flex.Flex,
        {
          as: import_framer_motion.motion.div,
          className: styles.suggestionTipsWrapper,
          columnGap: 16,
          rowGap: 8,
          justifyContent: "space-between",
          flexWrap: "wrap",
          variants: suggestionTipsVariants,
          initial: "enter",
          animate: "default",
          exit: "exit",
          transition,
          children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_Flex.Flex,
              {
                className: styles.suggestionTipsContent,
                style: { overflow: "hidden" },
                columnGap: 16,
                rowGap: 8,
                children: tips.map((tip) => {
                  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { maxWidth: "100%", gap: 4, children: [
                    /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Text.Text, { textStyle: "small-emphasized", children: [
                      tip.shortcut,
                      " "
                    ] }),
                    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                      import_TextEllipsis.TextEllipsis,
                      {
                        className: (0, import_clsx.default)(
                          (0, import_textStyle_css.textStyle)({ textStyle: "small" }),
                          styles.suggstionTipsEllipsis
                        ),
                        children: tip.text
                      }
                    )
                  ] }, tip.shortcut + tip.text);
                })
              }
            ),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_ExternalLink.ExternalLink,
              {
                href: "https://docs.dynatrace.com/docs/get-started/dynatrace-ui/ui-filter-field",
                className: (0, import_clsx.default)(
                  (0, import_textStyle_css.textStyle)({ textStyle: "small" }),
                  styles.docsLink
                ),
                "data-dt-name": "FilterSyntax",
                children: intl.formatMessage(import_messages.filterFieldMessages.docsLink)
              }
            )
          ]
        }
      )
    }
  );
};
