// @dynatrace/strato-components/filters/filter-field/suggestions/Suggestion.sty.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Suggestion_css_exports = {};
__export(Suggestion_css_exports, {
  defaultSuggestion: () => defaultSuggestion,
  icon: () => icon,
  matchableSuggestionDetails: () => matchableSuggestionDetails,
  pastingSuggestion: () => pastingSuggestion,
  searchSuggestion: () => searchSuggestion,
  statementSuggestionValue: () => statementSuggestionValue,
  suggestion: () => suggestion,
  suggestionContent: () => suggestionContent,
  suggestionText: () => suggestionText,
  suggestionsEmptyState: () => suggestionsEmptyState,
  suggestionsJSONPathHint: () => suggestionsJSONPathHint,
  suggestionsLoading: () => suggestionsLoading,
  tooltipWrapper: () => tooltipWrapper,
  truncateWithinMark: () => truncateWithinMark,
  value: () => value,
  virtualizedSuggestion: () => virtualizedSuggestion
});
module.exports = __toCommonJS(Suggestion_css_exports);
var import_SuggestionsOverlay_css_ts_vanilla = require("./SuggestionsOverlay.css");
var import_AppRoot_css_ts_vanilla = require("../../../core/components/app-root/AppRoot.css");
var import_virtualFocus_css_ts_vanilla = require("../../../core/styles/virtualFocus.css");
var import_Suggestion_css_ts_vanilla = require("./Suggestion.css");
var defaultSuggestion = "_1smjoa1c-3-6-0";
var icon = "_1smjoa10-3-6-0";
var matchableSuggestionDetails = "_1smjoa12-3-6-0";
var pastingSuggestion = "_1smjoa1f-3-6-0";
var searchSuggestion = "_1smjoa1d-3-6-0";
var statementSuggestionValue = "_1smjoa1e-3-6-0";
var suggestion = "_1smjoa14-3-6-0";
var suggestionContent = "_1smjoa15-3-6-0";
var suggestionText = "_1smjoa16-3-6-0";
var suggestionsEmptyState = "_1smjoa17-3-6-0";
var suggestionsJSONPathHint = "_1smjoa18-3-6-0";
var suggestionsLoading = "_1smjoa19-3-6-0";
var tooltipWrapper = "_1smjoa1b-3-6-0";
var truncateWithinMark = "_1smjoa1a-3-6-0";
var value = "_1smjoa11-3-6-0";
var virtualizedSuggestion = "_1smjoa13-3-6-0";
