// @dynatrace/strato-components/filters/filter-field/suggestions/SuggestionsOverlay.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SuggestionsOverlay_exports = {};
__export(SuggestionsOverlay_exports, {
  SuggestionsOverlay: () => SuggestionsOverlay
});
module.exports = __toCommonJS(SuggestionsOverlay_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_framer_motion = require("framer-motion");
var import_react = require("react");
var styles = __toESM(require("./SuggestionsOverlay.sty.js"));
var import_SuggestionTips = require("./SuggestionTips.js");
var import_OverlayContainer = require("../../../core/components/OverlayContainer.js");
var import_useOverlayStacking = require("../../../core/hooks/useOverlayStacking.js");
var import_OverlayStackingProvider = require("../../../core/providers/OverlayStackingProvider.js");
var import_useOverlayTransition = __toESM(require("../hooks/useOverlayTransition.js"));
const suggestionsOverlayAnimationVariants = {
  inital: {
    opacity: 0.4,
    scaleY: 0.8
  },
  default: {
    opacity: 1,
    scaleY: 1
  },
  exit: {
    opacity: 0,
    scaleY: 0.8
  }
};
const SuggestionsOverlay = (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    children,
    overlayProps,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow
  } = props;
  const overlayContainerProps = (0, import_useOverlayStacking.useOverlayStacking)(void 0, "dropdown");
  const transition = (0, import_useOverlayTransition.default)(false);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_OverlayContainer.OverlayContainer,
    {
      className: "suggestions-overlay",
      ...overlayContainerProps,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayStackingProvider.OverlayStackingProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        "div",
        {
          ref: forwardedRef,
          "data-testid": dataTestId,
          onPointerDown: (event) => {
            event.preventDefault();
          },
          ...overlayProps,
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
            import_framer_motion.motion.div,
            {
              className: styles.suggestionsOverlayWrapper,
              variants: suggestionsOverlayAnimationVariants,
              initial: "inital",
              animate: "default",
              exit: "exit",
              transition,
              children: [
                children,
                /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_SuggestionTips.SuggestionTips, {})
              ]
            }
          )
        }
      ) })
    }
  );
});
