// @dynatrace/strato-components/filters/filter-field/suggestions/SuggestionPin.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SuggestionPin_exports = {};
__export(SuggestionPin_exports, {
  SuggestionPin: () => SuggestionPin,
  useSuggestionPin: () => useSuggestionPin
});
module.exports = __toCommonJS(SuggestionPin_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_Button = require("../../../buttons/button/Button.js");
var import_getVirtualFocusStyles = require("../../../core/styles/getVirtualFocusStyles.js");
var import_virtualFocus_css = require("../../../core/styles/virtualFocus.sty.js");
var import_Tooltip = require("../../../overlays/tooltip/Tooltip.js");
var import_FilterFieldContext = require("../contexts/FilterFieldContext.js");
var import_usePinnedFilters = require("../hooks/recent-and-pinned/usePinnedFilters.js");
var import_useRecentFiltersStore = require("../hooks/recent-and-pinned/useRecentFiltersStore.js");
var import_useFilterFieldContext = require("../hooks/useFilterFieldContext.js");
var import_messages = require("../messages.js");
function useSuggestionPin(filterString) {
  const { filterNamespace } = (0, import_react.useContext)(import_FilterFieldContext.FilterFieldContext);
  const pinnedFilters = (0, import_usePinnedFilters.usePinnedFilters)(filterNamespace);
  const { pinFilter, unpinFilter } = (0, import_usePinnedFilters.usePinnedFilterMutation)(filterNamespace);
  const isPinned = (pinnedFilters ?? []).find(
    (storedFilter) => storedFilter.filter === filterString.trim()
  ) !== void 0;
  return {
    isPinned,
    pinFilter,
    unpinFilter,
    limitReached: pinnedFilters.length >= 10
  };
}
function SuggestionPin(props) {
  const {
    id,
    filterString,
    className,
    size,
    placement,
    fallbackPlacements,
    role,
    disabled,
    "aria-disabled": ariaDisabledProps,
    showError,
    ...remainingProps
  } = props;
  const intl = (0, import_react_intl.useIntl)();
  const { focusedListItem, filterNamespace } = (0, import_useFilterFieldContext.useFilterFieldContext)();
  const canPin = (0, import_useRecentFiltersStore.useRecentAndPinnedFiltersStore)(
    (state) => state.canPin,
    filterNamespace
  );
  const { isPinned, pinFilter, unpinFilter, limitReached } = useSuggestionPin(filterString);
  const pinUnpinTooltip = intl.formatMessage(
    import_messages.filterFieldMessages.suggestions.pinUnpinTooltip,
    { isPinned }
  );
  const pinUnpinLabel = intl.formatMessage(
    import_messages.filterFieldMessages.suggestions.pinUnpinLabel,
    { isPinned, filterString }
  );
  const disabledTooltip = intl.formatMessage(
    import_messages.filterFieldMessages.suggestions.pinUnpinLimitTooltip
  );
  const ariaDisabled = limitReached && !isPinned;
  const buttonRef = (0, import_react.useRef)(null);
  const [tooltipOpen, setTooltipOpen] = (0, import_react.useState)(false);
  (0, import_react.useEffect)(() => {
    const classListObserver = new MutationObserver((entries) => {
      for (const entry of entries) {
        if (entry.target === buttonRef.current && entry.target instanceof HTMLElement) {
          if (entry.target.classList.contains(import_virtualFocus_css.virtualFocus.neutral)) {
            setTooltipOpen(true);
          } else {
            setTooltipOpen(false);
          }
        }
      }
    });
    if (buttonRef.current) {
      classListObserver.observe(buttonRef.current, {
        attributes: true,
        attributeFilter: ["class"]
      });
    }
    return () => {
      classListObserver.disconnect();
    };
  }, []);
  if (!canPin) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Tooltip.Tooltip,
    {
      open: !disabled && tooltipOpen,
      onOpenChange: setTooltipOpen,
      placement,
      fallbackPlacements,
      text: ariaDisabled ? disabledTooltip : pinUnpinTooltip,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_Button.Button,
        {
          size,
          role,
          ref: buttonRef,
          id,
          disabled,
          ...remainingProps,
          "data-suggestion-pin": true,
          ...role === "option" ? {
            "aria-selected": focusedListItem?.id === id,
            "aria-checked": isPinned
          } : { "aria-pressed": isPinned },
          "aria-label": ariaDisabled ? disabledTooltip : pinUnpinLabel,
          "aria-disabled": ariaDisabledProps || ariaDisabled,
          color: showError ? "critical" : "neutral",
          className: (0, import_clsx.default)(
            className,
            (0, import_getVirtualFocusStyles.getVirtualFocusStyles)(focusedListItem?.id === id).className
          ),
          onPointerDown: (event) => {
            event.preventDefault();
            event.stopPropagation();
            event.nativeEvent.stopImmediatePropagation();
          },
          onClick: () => {
            if (ariaDisabled) {
              return;
            }
            if (isPinned) {
              unpinFilter(filterString);
            } else {
              pinFilter(filterString);
            }
          },
          children: isPinned ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.PinOffIcon, {}) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.PinIcon, {})
        }
      )
    }
  );
}
