// @dynatrace/strato-components/esm/notifications/toast/ToastContainer.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import {
  forwardRef,
  useCallback,
  useEffect,
  useRef
} from "react";
import { createPortal } from "react-dom";
import { useIntl } from "react-intl";
import {
  ToastContainer as ReactToastifyContainer,
  cssTransition,
  toast
} from "react-toastify";
import { useSyncExternalStoreWithSelector } from "use-sync-external-store/shim/with-selector";
import { XmarkIcon } from "@dynatrace/strato-icons";
import * as styles from "./ToastContainer.sty.js";
import { Button } from "../../buttons/button/Button.js";
import { useAriaLabelingProps } from "../../core/hooks/useAriaLabelingProps.js";
import { useMergeRefs } from "../../core/hooks/useMergeRefs.js";
import { useSafeForwardProps } from "../../core/hooks/useSafeForwardProps.js";
import { createStore } from "../../core/store/store.js";
import { Flex } from "../../layouts/flex/Flex.js";
import { useKeyboardFocusHandling } from "../../overlays/hooks/useKeyboardFocusHandling.js";
const COMPONENT_NAME = "ToastContainer";
let unique = 0;
const toastContainerStore = createStore({
  currentContainerId: null
});
const ToastContainer = forwardRef(
  (props, forwardedRef) => {
    const {
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...remainingProps
    } = props;
    const closeButtonRef = useRef(null);
    const containerRef = useRef(null);
    const ariaLabelingProps = useAriaLabelingProps(remainingProps);
    const forwardProps = useSafeForwardProps(remainingProps, ariaLabelingProps);
    const { "aria-label": ariaLabel, ...remainingAriaLabelingProps } = ariaLabelingProps;
    const [handleFocusOnKeydown, closedOnKeydown] = useKeyboardFocusHandling(
      closeButtonRef,
      containerRef
    );
    const rootRef = useMergeRefs([forwardedRef, containerRef]);
    const CloseButton = ({
      closeToast
    }) => {
      const intl = useIntl();
      const closeButtonLabel = intl.formatMessage({
        defaultMessage: "Close notification",
        id: "7IzOa/lx4Dc/qLGo",
        description: "Aria-label for the close notification button."
      });
      const closeToastOnEsc = (event) => {
        if (event.key === "Escape") {
          event.preventDefault();
          closedOnKeydown.current = true;
          closeToast(event);
        }
      };
      const detectCloseOnKeyDown = useCallback(() => {
        closedOnKeydown.current = true;
      }, []);
      return /* @__PURE__ */ jsx(
        Flex,
        {
          flexDirection: "column",
          alignItems: "flex-start",
          paddingTop: 6,
          paddingRight: 6,
          onKeyDown: closeToastOnEsc,
          children: /* @__PURE__ */ jsx(
            Button,
            {
              onKeyDown: detectCloseOnKeyDown,
              onClick: closeToast,
              "aria-label": closeButtonLabel,
              ref: closeButtonRef,
              variant: "default",
              onPointerDown: (e) => e.stopPropagation(),
              "data-dt-name": "CloseX",
              children: /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(XmarkIcon, {}) })
            }
          )
        }
      );
    };
    const containerId = useRef(++unique);
    const activeId = useSyncExternalStoreWithSelector(
      toastContainerStore.subscribe,
      () => toastContainerStore.getState(),
      null,
      (state) => state.currentContainerId
    );
    useEffect(() => {
      const id = containerId.current;
      if (toastContainerStore.getState().currentContainerId === null) {
        toastContainerStore.setState({ currentContainerId: id });
      }
    }, [activeId]);
    useEffect(() => {
      const id = containerId.current;
      let pendingObserver = null;
      const unsubscribeOnChange = toast.onChange((payload) => {
        if (payload.status === "removed") {
          const container = containerRef.current;
          if (!container) {
            return;
          }
          pendingObserver?.disconnect();
          pendingObserver = new MutationObserver((mutations) => {
            for (const mutation of mutations) {
              if (mutation.removedNodes.length > 0) {
                pendingObserver?.disconnect();
                pendingObserver = null;
                handleFocusOnKeydown();
                return;
              }
            }
          });
          pendingObserver.observe(container, {
            childList: true,
            subtree: true
          });
        }
      });
      return function cleanup() {
        unsubscribeOnChange();
        pendingObserver?.disconnect();
        if (toastContainerStore.getState().currentContainerId === id) {
          toastContainerStore.setState({ currentContainerId: null });
        }
      };
    }, []);
    if (activeId !== containerId.current) {
      return null;
    }
    return createPortal(
      /* @__PURE__ */ jsx(
        "div",
        {
          ref: rootRef,
          className: clsx(consumerClassName, styles.root),
          style: consumerStyle,
          "data-toast-container": true,
          "data-testid": dataTestId,
          "data-dtrum-mask": dataDtrumMask,
          "data-dtrum-allow": dataDtrumAllow,
          ...remainingAriaLabelingProps,
          ...forwardProps,
          "data-dt-component": COMPONENT_NAME,
          children: /* @__PURE__ */ jsx(
            ReactToastifyContainer,
            {
              hideProgressBar: true,
              draggable: false,
              closeButton: CloseButton,
              limit: 5,
              pauseOnHover: true,
              closeOnClick: false,
              newestOnTop: false,
              ...ariaLabel && { "aria-label": ariaLabel },
              transition: cssTransition({
                collapseDuration: 200,
                enter: "strato__toast__enter",
                exit: "strato__toast__exit"
              })
            }
          )
        }
      ),
      document.body
    );
  }
);
ToastContainer.displayName = COMPONENT_NAME;
export {
  ToastContainer
};
