// @dynatrace/strato-components/esm/notifications/toast/Toast.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import {
  useMemo
} from "react";
import { toast } from "react-toastify";
import {
  CriticalIcon,
  InformationIcon,
  SuccessIcon,
  WarningIcon
} from "@dynatrace/strato-icons";
import {
  toastCSS,
  toastFlexContainerCSS,
  toastIconCSS,
  toastTextCSS
} from "./Toast.sty.js";
import { OverlayStackingContext } from "../../core/contexts/OverlayStacking.js";
import { useAriaLabelingProps } from "../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { Container } from "../../layouts/container/Container.js";
import { Flex } from "../../layouts/flex/Flex.js";
import { textStyle as textStyleCSS } from "../../styles/textStyle.sty.js";
import { Text } from "../../typography/text/Text.js";
const COMPONENT_NAME = "Toast";
const Toast = (props) => {
  const {
    children,
    type,
    title,
    message,
    closeToast,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
  const ariaLabelingProps = useAriaLabelingProps(restProps);
  const Icon = useMemo(() => {
    switch (type) {
      case "warning":
        return /* @__PURE__ */ jsx(WarningIcon, {});
      case "success":
        return /* @__PURE__ */ jsx(SuccessIcon, {});
      case "critical":
        return /* @__PURE__ */ jsx(CriticalIcon, {});
      case "info":
        return /* @__PURE__ */ jsx(InformationIcon, {});
    }
  }, [type]);
  const closeToastOnEsc = (e) => {
    if (e.key === "Escape") {
      e.preventDefault();
      closeToast();
    }
  };
  return /* @__PURE__ */ jsx(
    "div",
    {
      onKeyDown: closeToastOnEsc,
      className: clsx(consumerClassName, toastCSS),
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...ariaLabelingProps,
      ...behavioralTrackingProps,
      children: /* @__PURE__ */ jsxs(OverlayStackingContext.Provider, { value: 100, children: [
        /* @__PURE__ */ jsx(
          Container,
          {
            variant: "accent",
            color: type === "info" ? "primary" : type,
            className: toastIconCSS,
            children: Icon
          }
        ),
        /* @__PURE__ */ jsxs(
          Flex,
          {
            className: clsx(
              toastFlexContainerCSS,
              textStyleCSS({ textStyle: "base" })
            ),
            paddingLeft: 16,
            paddingTop: 12,
            paddingBottom: 12,
            paddingRight: 8,
            flexDirection: "column",
            alignItems: "flex-start",
            flexGrow: 1,
            gap: 6,
            children: [
              title && /* @__PURE__ */ jsx(Text, { className: toastTextCSS, textStyle: "base-emphasized", as: "p", children: title }),
              message && /* @__PURE__ */ jsx(Text, { className: toastTextCSS, textStyle: "base", as: "p", children: message }),
              children
            ]
          }
        )
      ] })
    }
  );
};
function showToast({
  id,
  type = "info",
  role,
  lifespan,
  title,
  message,
  position = "bottom-left",
  actions,
  className: consumerClassName,
  style: consumerStyle,
  "aria-label": ariaLabel,
  "aria-labelledby": ariaLabelledBy,
  "aria-describedby": ariaDescribedBy,
  "aria-details": ariaDetails,
  "data-testid": dataTestId,
  "data-dtrum-mask": dataDtrumMask,
  "data-dtrum-allow": dataDtrumAllow,
  ...remainingProps
}) {
  const defaultRole = type === "critical" ? "alert" : "status";
  const defaultAutoClose = type === "warning" || type === "critical" ? false : 8e3;
  const autoClose = lifespan === "infinite" ? false : lifespan;
  return toast(
    /* @__PURE__ */ jsx(
      Toast,
      {
        type,
        title,
        message,
        closeToast: () => void 0,
        className: consumerClassName,
        style: consumerStyle,
        "aria-label": ariaLabel,
        "aria-labelledby": ariaLabelledBy,
        "aria-describedby": ariaDescribedBy,
        "aria-details": ariaDetails,
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        ...remainingProps,
        "data-dt-component": COMPONENT_NAME,
        children: actions
      }
    ),
    {
      toastId: id,
      autoClose: autoClose ?? defaultAutoClose,
      role: role ?? defaultRole,
      position
    }
  );
}
function dismissToast(id) {
  toast.dismiss(id);
}
function dismissAllToasts() {
  toast.dismiss();
}
export {
  dismissAllToasts,
  dismissToast,
  showToast
};
