// @dynatrace/strato-components/charts/core/components/chart-layout/types/constants.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var constants_exports = {};
__export(constants_exports, {
  DEFAULT_LAYOUT_HEIGHT: () => DEFAULT_LAYOUT_HEIGHT,
  DEFAULT_LAYOUT_WIDTH: () => DEFAULT_LAYOUT_WIDTH,
  DEFAULT_LEGEND_CONFIG: () => DEFAULT_LEGEND_CONFIG,
  DEFAULT_POSITION: () => DEFAULT_POSITION,
  DEFAULT_RATIO_BOUNDARIES: () => DEFAULT_RATIO_BOUNDARIES,
  DEFAULT_RESIZER_SIZE: () => DEFAULT_RESIZER_SIZE,
  LOADING_STATE_OPACITY: () => LOADING_STATE_OPACITY,
  MAX_AUTO_RATIO_SIZE: () => MAX_AUTO_RATIO_SIZE,
  MIN_LEGEND_HEIGHT: () => MIN_LEGEND_HEIGHT,
  MIN_LEGEND_WIDTH: () => MIN_LEGEND_WIDTH,
  MIN_PLOT_SIZE: () => MIN_PLOT_SIZE,
  TEMPLATE_AREA_FOR_FIXED_CHART_LAYOUT: () => TEMPLATE_AREA_FOR_FIXED_CHART_LAYOUT,
  TEMPLATE_AREA_FOR_GRAPH: () => TEMPLATE_AREA_FOR_GRAPH,
  TEMPLATE_AREA_FOR_RESIZABLE_CHART_LAYOUT: () => TEMPLATE_AREA_FOR_RESIZABLE_CHART_LAYOUT
});
module.exports = __toCommonJS(constants_exports);
var import_constants = require("../../../constants.js");
const DEFAULT_POSITION = "auto";
const DEFAULT_RATIO_BOUNDARIES = {
  minWidth: 0.1,
  minHeight: 0.1,
  maxWidth: 0.8,
  maxHeight: 0.7
};
const DEFAULT_LAYOUT_HEIGHT = "300";
const DEFAULT_LAYOUT_WIDTH = "100%";
const DEFAULT_LEGEND_CONFIG = {
  position: DEFAULT_POSITION,
  ratioBoundaries: DEFAULT_RATIO_BOUNDARIES,
  ratio: import_constants.defaultRatio,
  overrideRatio: void 0
};
const LOADING_STATE_OPACITY = 40;
const DEFAULT_RESIZER_SIZE = 15;
const MAX_AUTO_RATIO_SIZE = 0.3;
const MIN_LEGEND_HEIGHT = 30;
const MIN_LEGEND_WIDTH = 70;
const MIN_PLOT_SIZE = 100;
const TEMPLATE_AREA_FOR_GRAPH = '"graphArea"';
const TEMPLATE_AREA_FOR_RESIZABLE_CHART_LAYOUT = '"graphArea horizontalResize verticalLegend" "verticalResize . ." "horizontalLegend horizontalLegend horizontalLegend"';
const TEMPLATE_AREA_FOR_FIXED_CHART_LAYOUT = '"graphArea verticalLegend" "horizontalLegend horizontalLegend"';
