// @dynatrace/strato-components/esm/content/feature-highlight/FeatureHighlight.js
import { jsx, jsxs } from "react/jsx-runtime";
import { eventListeners } from "@popperjs/core";
import {
  forwardRef,
  useEffect,
  useLayoutEffect,
  useMemo,
  useRef,
  useState,
  Children,
  isValidElement
} from "react";
import { createPortal } from "react-dom";
import { useIntl } from "react-intl";
import { usePopper } from "react-popper";
import { XmarkIcon } from "@dynatrace/strato-icons";
import { Actions } from "./Actions.js";
import {
  FeatureHighlightCloseButtonCSS,
  FeatureHighlightPointerArrowCSS,
  FeatureHighlightScrollContainerCSS
} from "./FeatureHighlight.sty.js";
import { Button } from "../../buttons/button/Button.js";
import { useAriaLabelingProps } from "../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useMergeRefs } from "../../core/hooks/useMergeRefs.js";
import { useOverlayStacking } from "../../core/hooks/useOverlayStacking.js";
import { useSafeForwardProps } from "../../core/hooks/useSafeForwardProps.js";
import { isSlot } from "../../core/utils/is-slot.js";
import { devLog } from "../../core/utils/logging.js";
import { Flex } from "../../layouts/flex/Flex.js";
import { useKeyboardFocusHandling } from "../../overlays/hooks/useKeyboardFocusHandling.js";
import {
  basePlacements
} from "../../overlays/types/placement.js";
import {
  observeAnimationDone,
  observeOverlaySize
} from "../../overlays/utils/popper-modifiers.js";
import { Content } from "../shared-components/Content.js";
import { Title } from "../shared-components/Title.js";
import { Visual } from "../shared-components/Visual.js";
const COMPONENT_NAME = "FeatureHighlight";
const getPreferredPlacement = (placement) => {
  const preferredPlacement = placement.split("-")[0];
  if (basePlacements.includes(preferredPlacement)) {
    return preferredPlacement;
  } else {
    return basePlacements[1];
  }
};
const FeatureHighlightRoot = forwardRef(
  (props, forwardedRef) => {
    const {
      open,
      anchor,
      placement = "right-bottom",
      onClose,
      children,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      className: consumerClassName,
      style: consumerStyle,
      ...remainingProps
    } = props;
    const preferredPlacement = getPreferredPlacement(placement);
    const [overlayElement, setOverlayElement] = useState();
    const [arrowElement, setArrowElement] = useState();
    const closeButtonRef = useRef(null);
    const featureHighlightContainer = useRef(null);
    const rootRef = useMergeRefs([
      setOverlayElement,
      forwardedRef
    ]);
    const [handleFocusOnKeydown, closedOnKeydown] = useKeyboardFocusHandling(closeButtonRef);
    const overlayStacking = useOverlayStacking(void 0, "feature-highlight");
    const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
    const ariaLabelingProps = useAriaLabelingProps(restProps);
    const forwardProps = useSafeForwardProps(restProps, ariaLabelingProps);
    const intl = useIntl();
    const slots = {
      title: null,
      visual: null,
      content: null,
      actions: null
    };
    useLayoutEffect(() => {
      const toastContainer = document.querySelector("[data-toast-container]");
      const microguideContainer = document.querySelector(
        "[data-microguide-container]"
      );
      featureHighlightContainer.current = document.createElement("div");
      featureHighlightContainer.current.setAttribute("aria-live", "polite");
      featureHighlightContainer.current.setAttribute(
        "data-feature-highlight-container",
        "true"
      );
      if (microguideContainer && microguideContainer.parentNode) {
        microguideContainer.parentNode.insertBefore(
          featureHighlightContainer.current,
          microguideContainer
        );
      } else if (toastContainer && toastContainer.parentNode) {
        toastContainer.parentNode.insertBefore(
          featureHighlightContainer.current,
          toastContainer
        );
      } else {
        document.body.appendChild(featureHighlightContainer.current);
      }
      return () => {
        featureHighlightContainer.current?.remove();
      };
    }, []);
    Children.forEach(children, (child) => {
      if (isSlot(child, Title)) {
        slots.title = child;
      } else if (isSlot(child, Visual)) {
        slots.visual = child;
      } else if (isSlot(child, Content)) {
        slots.content = child;
      } else if (isSlot(child, Actions)) {
        slots.actions = child;
      } else if (isValidElement(child)) {
        devLog(
          "Only Title, Visual, Content and Actions are valid children of FeatureHighlight."
        );
      } else if (child !== null) {
        devLog("The used child is not supported by FeatureHighlight.");
      }
    });
    if (slots.title && ariaLabelingProps["aria-labelledby"] === void 0) {
      ariaLabelingProps["aria-labelledby"] = "strato-feature-highlight-title";
    }
    if (slots.content && ariaLabelingProps["aria-describedby"] === void 0) {
      ariaLabelingProps["aria-describedby"] = "strato-feature-highlight-content";
    }
    const closeOnEscape = (e) => {
      if (e.key === "Escape") {
        e.preventDefault();
        closedOnKeydown.current = true;
        onClose();
      }
    };
    const offsetModifier = useMemo(
      () => ({
        name: "offset",
        options: {
          offset: (({ placement: popperPlacement, popper }) => {
            let skidding = 0;
            const spacingToPopperEdge = 24;
            if (popperPlacement === preferredPlacement) {
              switch (placement) {
                case "right-bottom":
                case "left-bottom":
                  skidding = popper.height / 2 - spacingToPopperEdge;
                  break;
                case "right-top":
                case "left-top":
                  skidding = -(popper.height / 2 - spacingToPopperEdge);
                  break;
                case "top-left":
                case "bottom-left":
                  skidding = -(popper.width / 2 - spacingToPopperEdge);
                  break;
                case "top-right":
                case "bottom-right":
                  skidding = popper.width / 2 - spacingToPopperEdge;
                  break;
                default:
                  skidding = 0;
              }
            }
            return [skidding, 16];
          })
        }
      }),
      [placement, preferredPlacement]
    );
    const { styles, attributes, update } = usePopper(anchor, overlayElement, {
      placement: preferredPlacement,
      modifiers: [
        offsetModifier,
        {
          name: "flip",
          options: {
            fallbackPlacements: [...basePlacements]
          }
        },
        {
          name: "arrow",
          options: {
            element: arrowElement,
            padding: 16
          }
        },
        observeOverlaySize,
        observeAnimationDone,
        eventListeners
      ]
    });
    useEffect(() => {
      if (!anchor || !open || !update) {
        return;
      }
      let prevRect = anchor.getBoundingClientRect();
      let positionPollRequestId = null;
      let pendingAnimationFrame = null;
      let isUpdateScheduled = false;
      const scheduleUpdate = () => {
        if (isUpdateScheduled) {
          return;
        }
        isUpdateScheduled = true;
        if (pendingAnimationFrame !== null) {
          cancelAnimationFrame(pendingAnimationFrame);
        }
        pendingAnimationFrame = requestAnimationFrame(async () => {
          await update();
          isUpdateScheduled = false;
          pendingAnimationFrame = null;
        });
      };
      const pollForPositionChange = () => {
        const anchorRect = anchor.getBoundingClientRect();
        const hasPositionChanged = Math.abs(anchorRect.x - prevRect.x) > 0.5 || Math.abs(anchorRect.y - prevRect.y) > 0.5 || Math.abs(anchorRect.width - prevRect.width) > 0.5 || Math.abs(anchorRect.height - prevRect.height) > 0.5;
        if (hasPositionChanged) {
          prevRect = anchorRect;
          scheduleUpdate();
        }
        positionPollRequestId = requestAnimationFrame(pollForPositionChange);
      };
      positionPollRequestId = requestAnimationFrame(pollForPositionChange);
      return () => {
        if (positionPollRequestId !== null) {
          cancelAnimationFrame(positionPollRequestId);
        }
        if (pendingAnimationFrame !== null) {
          cancelAnimationFrame(pendingAnimationFrame);
        }
      };
    }, [anchor, open, update]);
    useEffect(() => {
      if (open === false) {
        handleFocusOnKeydown();
      }
    }, [handleFocusOnKeydown, open]);
    return open && anchor ? createPortal(
      /* @__PURE__ */ jsxs(
        Flex,
        {
          ref: rootRef,
          onKeyDown: closeOnEscape,
          role: "dialog",
          "aria-modal": "false",
          className: consumerClassName,
          style: {
            ...styles.popper,
            zIndex: overlayStacking.style.zIndex,
            ...consumerStyle
          },
          "data-testid": dataTestId,
          "data-dtrum-mask": dataDtrumMask,
          "data-dtrum-allow": dataDtrumAllow,
          ...attributes.popper,
          ...ariaLabelingProps,
          ...behavioralTrackingProps,
          ...forwardProps,
          "data-dt-component": COMPONENT_NAME,
          children: [
            /* @__PURE__ */ jsxs(
              Flex,
              {
                className: FeatureHighlightScrollContainerCSS,
                flexDirection: "column",
                gap: 0,
                paddingX: 16,
                paddingTop: 8,
                paddingBottom: 8,
                children: [
                  /* @__PURE__ */ jsxs(
                    Flex,
                    {
                      flexDirection: "row",
                      justifyContent: "space-between",
                      alignItems: "flex-start",
                      gap: 8,
                      paddingBottom: slots.content !== null || slots.actions !== null || slots.visual !== null ? 8 : 0,
                      children: [
                        /* @__PURE__ */ jsx(
                          Flex,
                          {
                            flexItem: true,
                            flexGrow: 1,
                            paddingY: 6,
                            id: "strato-feature-highlight-title",
                            children: slots.title
                          }
                        ),
                        /* @__PURE__ */ jsx(
                          Button,
                          {
                            className: FeatureHighlightCloseButtonCSS,
                            onClick: () => onClose(),
                            onKeyDown: () => closedOnKeydown.current = true,
                            ref: closeButtonRef,
                            "aria-label": intl.formatMessage({
                              defaultMessage: "Close feature highlight",
                              id: "XQHEJyt2t/js0TGr",
                              description: "Close the feature highlight."
                            }),
                            "data-dt-name": "CloseX",
                            children: /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(XmarkIcon, {}) })
                          }
                        )
                      ]
                    }
                  ),
                  slots.visual && /* @__PURE__ */ jsx(
                    Flex,
                    {
                      paddingBottom: slots.content !== null || slots.actions !== null ? 16 : 8,
                      children: slots.visual
                    }
                  ),
                  slots.content && /* @__PURE__ */ jsx(
                    Flex,
                    {
                      paddingBottom: 8,
                      style: { overflow: "hidden" },
                      id: "strato-feature-highlight-content",
                      children: slots.content
                    }
                  ),
                  slots.actions
                ]
              }
            ),
            /* @__PURE__ */ jsx(
              "div",
              {
                "data-popper-arrow": true,
                ref: setArrowElement,
                className: FeatureHighlightPointerArrowCSS,
                style: styles.arrow,
                ...attributes.arrow
              }
            )
          ]
        }
      ),
      featureHighlightContainer.current || document.body
    ) : null;
  }
);
FeatureHighlightRoot.displayName = COMPONENT_NAME;
const FeatureHighlight = Object.assign(FeatureHighlightRoot, {
  Title,
  Visual,
  Content,
  Actions
});
export {
  FeatureHighlight
};
