// @dynatrace/strato-components/charts/core/components/legend/range/components/RangeLegendItemMarker/RangeLegendItemMarker.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var RangeLegendItemMarker_exports = {};
__export(RangeLegendItemMarker_exports, {
  RangeLegendItemMarker: () => RangeLegendItemMarker
});
module.exports = __toCommonJS(RangeLegendItemMarker_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
var import_RangeLegendItemMarker_css = require("./RangeLegendItemMarker.sty.js");
var import_constants = require("../../../../../constants.js");
var import_constants2 = require("../../constants.js");
var import_RangeLegendOrientationContext = require("../../context/RangeLegendOrientationContext.js");
var import_RangeLegendScaleContext = require("../../context/RangeLegendScaleContext.js");
var import_selectors = require("../../hooks/store/selectors.js");
const RangeLegendItemMarker = (props) => {
  const { markerValue, position, isReversed, isSinglePointRange } = props;
  const svgRef = (0, import_react.useRef)(null);
  const hoveredMarkerId = (0, import_selectors.useRangeLegendHoveredMarkerId)();
  const { scale } = (0, import_RangeLegendScaleContext.useScale)();
  const isBoundarySegment = scale.range().includes(position);
  const { orientation } = (0, import_RangeLegendOrientationContext.useOrientation)();
  const isVertical = orientation === import_constants.ORIENTATION.Vertical;
  const isHorizontal = orientation === import_constants.ORIENTATION.Horizontal;
  const itemX = isHorizontal ? position + import_constants2.DEFAULT_GAP_THICKNESS / 2 : 0;
  const itemY = isVertical ? position + import_constants2.DEFAULT_GAP_THICKNESS / 2 : 0;
  const itemWidth = isHorizontal ? import_constants2.DEFAULT_GAP_THICKNESS : import_constants2.DEFAULT_ITEM_THICKNESS;
  const itemHeight = isVertical ? import_constants2.DEFAULT_GAP_THICKNESS : import_constants2.DEFAULT_ITEM_THICKNESS;
  const defaultTranslateY = `-${import_constants2.DEFAULT_GAP_THICKNESS / 2}`;
  const verticalSinglePointTranslateY = isReversed ? defaultTranslateY : `-${import_constants2.DEFAULT_GAP_THICKNESS * 2}`;
  const singlePointRangeTranslateY = isHorizontal ? import_constants2.DEFAULT_GAP_THICKNESS : verticalSinglePointTranslateY;
  const CSSTranslate = `translateX(${import_constants2.DEFAULT_GAP_THICKNESS}px) translateY(${isSinglePointRange && !isBoundarySegment ? singlePointRangeTranslateY : defaultTranslateY}px)`;
  const ITEM_X = {
    [import_constants.ORIENTATION.Vertical]: () => itemX - itemWidth + import_constants2.STROKE_WIDTH,
    [import_constants.ORIENTATION.Horizontal]: () => {
      if (isSinglePointRange && !isBoundarySegment) {
        return itemX + import_constants2.CONTAINER_PADDING - (import_constants2.DEFAULT_SEGMENT_SIZE + import_constants2.STROKE_WIDTH * 2) / 2;
      }
      return itemX + import_constants2.CONTAINER_PADDING - import_constants2.STROKE_WIDTH * 2;
    }
  };
  const ITEM_Y = {
    [import_constants.ORIENTATION.Vertical]: () => {
      if (isSinglePointRange && !isBoundarySegment) {
        return itemY - import_constants2.CONTAINER_PADDING - itemHeight / 2;
      }
      return itemY - import_constants2.CONTAINER_PADDING + import_constants2.STROKE_WIDTH;
    },
    [import_constants.ORIENTATION.Horizontal]: () => {
      return itemY - itemHeight + import_constants2.STROKE_WIDTH;
    }
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "svg",
    {
      width: "48",
      height: "13",
      viewBox: "0 0 48 13",
      fill: "none",
      ref: svgRef,
      x: ITEM_X[orientation](),
      y: ITEM_Y[orientation](),
      className: (0, import_RangeLegendItemMarker_css.legendMarker)({
        isMarkerShown: hoveredMarkerId === markerValue
      }),
      xmlns: "http://www.w3.org/2000/svg",
      role: "presentation",
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        "path",
        {
          d: "M7.27736 7.08243L7.90139 6.66639L7.27734 6.25038L1.27734 2.25065L0.5 1.73246L0.5 2.66668L0.5 10.6667L0.5 11.601L1.27736 11.0827L7.27736 7.08243Z",
          fill: import_colors.default.Icon.Neutral.Default,
          strokeWidth: import_constants2.STROKE_WIDTH,
          stroke: import_colors.default.Background.Base.Default,
          style: {
            transform: isHorizontal ? `rotate(90deg) ${CSSTranslate}` : CSSTranslate
          }
        }
      )
    }
  );
};
