// @dynatrace/strato-components/charts/xy-chart/components/axes/XYChartBottomAxisArea.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var XYChartBottomAxisArea_exports = {};
__export(XYChartBottomAxisArea_exports, {
  XYChartBottomAxisArea: () => XYChartBottomAxisArea
});
module.exports = __toCommonJS(XYChartBottomAxisArea_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("@use-gesture/react");
var import_clsx = __toESM(require("clsx"));
var import_react2 = require("react");
var import_AxisArea_css = require("./AxisArea.sty.js");
var import_compute_collision_aware_max_space = require("./utils/compute-collision-aware-max-space.js");
var import_is_axis_pannable = require("./utils/is-axis-pannable.js");
var import_XTickWithTruncation = require("./XTickWithTruncation.js");
var import_XYChartPanningControls = require("./XYChartPanningControls.js");
var import_AxisBottom = require("../../../core/components/axes/AxisBottom.js");
var import_DivAxisWrapper = require("../../../core/components/axes/components/DivAxisWrapper.js");
var import_constants = require("../../../core/components/axes/constants.js");
var import_calculate_axis_size = require("../../../core/components/axes/utils/calculate-axis-size.js");
var import_get_tick_label_props = require("../../../core/components/axes/utils/get-tick-label-props.js");
var import_is_tick_label_visible = require("../../../core/components/axes/utils/is-tick-label-visible.js");
var import_ChartGridRoot = require("../../../core/components/chart-grid/ChartGridRoot.js");
var import_constants2 = require("../../../core/components/zoom/constants.js");
var import_ZoomCursors_css = require("../../../core/components/zoom/ZoomCursors.sty.js");
var import_defaults = require("../../constants/defaults.js");
var import_useAxesTicks = require("../../hooks/useAxesTicks.js");
var import_useVisibleAxesConfig = require("../../hooks/useVisibleAxesConfig.js");
var import_useVisibleAxesScales = require("../../hooks/useVisibleAxesScales.js");
var import_useXYChartScales = require("../../hooks/useXYChartScales.js");
var import_guards = require("../../types/guards.js");
var import_grid_templates = require("../../utils/grid-templates.js");
var import_useApplyPanDelta = require("../overlay/hooks/useApplyPanDelta.js");
function XYChartBottomAxisArea() {
  const axisRef = (0, import_react2.useRef)(null);
  const { bottomTicks } = (0, import_useAxesTicks.useAxesTicks)();
  const { bottom: bottomConfig } = (0, import_useVisibleAxesConfig.useVisibleAxesConfig)();
  const { bottom: bottomScale } = (0, import_useVisibleAxesScales.useVisibleAxesScales)();
  const scales = (0, import_useXYChartScales.useXYChartScales)();
  const canPan = (0, import_is_axis_pannable.isAxisPannable)(bottomConfig, scales);
  const tickFormat = (_, index) => bottomTicks.formattedTicks[index];
  const isCategoricalAxis = bottomConfig?.type ? (0, import_guards.isCategorical)(bottomConfig.type) : false;
  const isTickRotated = bottomConfig?.tickLabelLayout === "vertical";
  const tickLabelProps = (0, import_get_tick_label_props.getXAxisTickLabelProps)(
    "bottom",
    isCategoricalAxis,
    isTickRotated,
    bottomScale,
    bottomConfig?.tickLabelLayout
  );
  const areCategoricalTicksNotFittable = bottomScale && isCategoricalAxis && !(0, import_is_tick_label_visible.isTickLabelSpaceAvailable)(bottomScale.step());
  const handlePan = (0, import_useApplyPanDelta.useApplyPanDelta)();
  (0, import_react.useGesture)(
    {
      onDrag: (e) => import_constants2.KEYBOARD_HANDLERS.includes(e.type) ? null : handlePan(-e.delta[0])
    },
    {
      target: axisRef,
      eventOptions: { passive: false },
      drag: { filterTaps: true, pointer: { buttons: [1, 4] } },
      enabled: canPan
    }
  );
  const hideAxis = bottomConfig === void 0 || areCategoricalTicksNotFittable || bottomScale === void 0;
  const tickNameHasEmoji = bottomTicks.ticks.some(
    (name) => import_constants.EMOJI_REGEX.test(name)
  );
  return hideAxis ? null : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ChartGridRoot.ChartGrid.GridArea, { gridArea: import_grid_templates.AREAS.bottomAxis, children: ({ availableWidth, availableHeight }) => {
    const maxSpaceForTruncation = (0, import_compute_collision_aware_max_space.computeCollisionAwareMaxSpace)(
      (0, import_calculate_axis_size.getXTickMaxSize)(
        availableHeight,
        availableWidth,
        isTickRotated,
        bottomScale,
        bottomConfig.label,
        bottomTicks.ticks.length
      ),
      bottomTicks.formattedTicks,
      bottomTicks.ticks,
      bottomScale,
      isCategoricalAxis,
      isTickRotated
    );
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      "div",
      {
        className: (0, import_clsx.default)(
          (0, import_ZoomCursors_css.cursorStyle)({ isPanEnabled: canPan }),
          import_AxisArea_css.axisContainerStyle
        ),
        ref: axisRef,
        children: [
          canPan && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_XYChartPanningControls.XYChartPanningControls,
            {
              visibleAxis: bottomConfig,
              onClickPan: (delta) => handlePan(-delta)
            }
          ),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_DivAxisWrapper.DivAxisWrapper,
            {
              divAxisStyle: import_constants.axisFontProps,
              ticks: bottomTicks.formattedTicks,
              label: bottomConfig.label,
              isTickVertical: true,
              isTickRotated,
              children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_AxisBottom.AxisBottom,
                {
                  role: "group",
                  "aria-label": import_defaults.XY_CHART_BOTTOM_AXIS_LABEL,
                  scale: bottomScale,
                  containerSize: availableWidth,
                  height: availableHeight,
                  tickValues: bottomTicks.ticks,
                  tickLabelProps,
                  tickFormat,
                  label: bottomConfig.label,
                  tickComponent: (renderProps) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                    import_XTickWithTruncation.XTickWithTruncation,
                    {
                      renderProps,
                      maxSpaceForTruncation,
                      centerTick: isCategoricalAxis,
                      isTickRotated,
                      position: "bottom",
                      tickNameHasEmoji
                    }
                  )
                }
              )
            }
          )
        ]
      }
    );
  } });
}
XYChartBottomAxisArea["displayName"] = "XYChartValueBottomAxisArea";
