// @dynatrace/strato-components/charts/xy-chart/components/axes/XYChartLeftAxisArea.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var XYChartLeftAxisArea_exports = {};
__export(XYChartLeftAxisArea_exports, {
  XYChartLeftAxisArea: () => XYChartLeftAxisArea
});
module.exports = __toCommonJS(XYChartLeftAxisArea_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_YTickWithTruncation = require("./YTickWithTruncation.js");
var import_AxisLeft = require("../../../core/components/axes/AxisLeft.js");
var import_DivAxisWrapper = require("../../../core/components/axes/components/DivAxisWrapper.js");
var import_constants = require("../../../core/components/axes/constants.js");
var import_calculate_axis_size = require("../../../core/components/axes/utils/calculate-axis-size.js");
var import_get_tick_label_props = require("../../../core/components/axes/utils/get-tick-label-props.js");
var import_is_tick_label_visible = require("../../../core/components/axes/utils/is-tick-label-visible.js");
var import_ChartGridRoot = require("../../../core/components/chart-grid/ChartGridRoot.js");
var import_useAxesTicks = require("../../hooks/useAxesTicks.js");
var import_useVisibleAxesConfig = require("../../hooks/useVisibleAxesConfig.js");
var import_useVisibleAxesScales = require("../../hooks/useVisibleAxesScales.js");
var import_guards = require("../../types/guards.js");
var import_grid_templates = require("../../utils/grid-templates.js");
const tickLabelProps = (0, import_get_tick_label_props.getTickLabelPropsForValueAxis)("left");
function XYChartLeftAxisArea() {
  const { leftTicks } = (0, import_useAxesTicks.useAxesTicks)();
  const { left: leftConfig } = (0, import_useVisibleAxesConfig.useVisibleAxesConfig)();
  const { left: leftScale } = (0, import_useVisibleAxesScales.useVisibleAxesScales)();
  const tickFormat = (_, index) => leftTicks.formattedTicks[index];
  const isCategoricalAxis = leftConfig?.type ? (0, import_guards.isCategorical)(leftConfig.type) : false;
  const areCategoricalTicksNotFittable = leftScale && isCategoricalAxis && !(0, import_is_tick_label_visible.isTickLabelSpaceAvailable)(leftScale.step());
  const hideAxis = leftConfig === void 0 || areCategoricalTicksNotFittable || leftScale === void 0;
  const tickNameHasEmoji = leftTicks.ticks.some(
    (name) => import_constants.EMOJI_REGEX.test(name)
  );
  return hideAxis ? null : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ChartGridRoot.ChartGrid.GridArea, { gridArea: import_grid_templates.AREAS.leftAxis, children: ({ availableWidth, availableHeight }) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_DivAxisWrapper.DivAxisWrapper,
    {
      divAxisStyle: import_constants.axisFontProps,
      ticks: leftTicks.formattedTicks,
      label: leftConfig.label,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_AxisLeft.AxisLeft,
        {
          scale: leftScale,
          containerSize: availableHeight,
          width: availableWidth,
          tickValues: leftTicks.ticks,
          tickLabelProps,
          tickFormat,
          label: leftConfig.label,
          tickComponent: (renderProps) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_YTickWithTruncation.YTickWithTruncation,
            {
              position: "left",
              renderProps,
              maxSpaceForTruncation: (0, import_calculate_axis_size.getYTickMaxSize)(
                availableWidth,
                leftConfig.label
              ),
              centerTick: isCategoricalAxis,
              tickNameHasEmoji
            }
          )
        }
      )
    }
  ) });
}
XYChartLeftAxisArea["displayName"] = "XYChartLeftAxisArea";
