// @dynatrace/strato-components/charts/meter-bar/single-meter-bar/components/MeterBarChartOverlay/MeterBarChartOverlayWithTooltip.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MeterBarChartOverlayWithTooltip_exports = {};
__export(MeterBarChartOverlayWithTooltip_exports, {
  MeterBarChartOverlayWithTooltip: () => MeterBarChartOverlayWithTooltip
});
module.exports = __toCommonJS(MeterBarChartOverlayWithTooltip_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_Flex = require("../../../../../layouts/flex/Flex.js");
var import_InteractionsOverlay_css = require("../../../../core/components/interactions-overlay/InteractionsOverlay.sty.js");
var import_utils = require("../../../../core/components/overlay-tooltip/_to-be-removed/utils.js");
var import_OverlayContext = require("../../../../core/components/overlay-tooltip/contexts/OverlayContext.js");
var import_OverlayTooltip = require("../../../../core/components/overlay-tooltip/OverlayTooltip.js");
var import_useTooltipActions = require("../../../../core/components/overlay-tooltip/useTooltipActions.js");
var import_utils2 = require("../../../../core/components/overlay-tooltip/utils.js");
var import_ChartSeriesActionContext = require("../../../../core/components/series-action/ChartSeriesActionContext.js");
var import_constants = require("../../../../core/constants.js");
var import_MeterBarChartTooltip = require("../../../shared/components/meter-bar-tooltip/MeterBarChartTooltip.js");
var import_FormatterContext = require("../../../shared/context/FormatterContext.js");
var import_MeterBarLayoutSizeContext = require("../../../shared/context/MeterBarLayoutSizeContext.js");
var import_element_size_css = require("../../../shared/styles/element-size.sty.js");
var import_styles_css = require("../../../shared/styles.sty.js");
var import_get_numeric_value = require("../../../shared/utils/get-numeric-value.js");
var import_MeterBarChartDataContext = require("../../context/MeterBarChartDataContext.js");
var import_MeterBarChartInternalConfigContext = require("../../context/MeterBarChartInternalConfigContext.js");
const MeterBarCustomTooltip = ({
  template,
  displayValue,
  min,
  max,
  color,
  name,
  thresholds
}) => {
  const activeThreshold = thresholds?.data?.slice().sort((a, b) => b.value - a.value).find((t) => displayValue >= t.value);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_OverlayTooltip.OverlayTooltip,
    {
      legacyTemplate: template({
        value: displayValue,
        min,
        max,
        color,
        name,
        threshold: activeThreshold ? { label: activeThreshold.name, color: activeThreshold.color } : void 0
      })
    }
  );
};
const MeterBarChartOverlayWithTooltip = (props) => {
  const { size } = (0, import_MeterBarLayoutSizeContext.useMeterBarLayoutSize)();
  const { color, name, value, valueAccessor } = (0, import_MeterBarChartDataContext.useMeterBarChartData)();
  const {
    tooltip: tooltipConfig,
    min,
    max,
    thresholds
  } = (0, import_MeterBarChartInternalConfigContext.useMeterBarChartInternalConfig)();
  const { formatter, unit } = (0, import_FormatterContext.useFormatter)();
  const { chartSeriesActions } = (0, import_ChartSeriesActionContext.useSeriesActions)();
  const overlay = (0, import_OverlayContext.useOverlay)();
  const { hover, togglePin, outOfBounds } = (0, import_useTooltipActions.useTooltipActions)();
  const { data: tooltipData } = (0, import_utils.useTooltipData)();
  const displayValue = (0, import_get_numeric_value.getNumericValue)(
    value,
    valueAccessor ?? import_constants.DEFAULT_VALUE_ACCESSOR
  );
  const item = {
    name: tooltipConfig?.name ?? name ?? "",
    value: formatter(displayValue, unit)
  };
  const data = {
    geometry: "meterbar",
    series: [
      {
        name: item.name,
        value: item.value,
        color
      }
    ]
  };
  const tooltipContent = tooltipConfig?.template && !(0, import_utils2.isTooltipDataEmpty)(tooltipData) ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    MeterBarCustomTooltip,
    {
      template: tooltipConfig.template,
      displayValue,
      min,
      max,
      color,
      name: tooltipConfig.name ?? item.name,
      thresholds
    }
  ) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_MeterBarChartTooltip.MeterBarChartTooltip,
    {
      actionsForItem: chartSeriesActions ? () => () => chartSeriesActions({
        id: `${name}-${displayValue}`,
        name: item.name,
        values: [item.value]
      }) : void 0
    }
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_Flex.Flex,
    {
      ref: overlay.overlayRef,
      className: (0, import_clsx.default)(import_styles_css.container, (0, import_element_size_css.elementSize)({ size }), import_InteractionsOverlay_css.interactionsOverlay),
      "data-testid": "overlay-wrapper",
      onPointerMove: (event) => hover(event, data),
      onPointerLeave: outOfBounds,
      onClick: (event) => togglePin(event, data),
      children: [
        tooltipContent,
        props.children
      ]
    }
  );
};
MeterBarChartOverlayWithTooltip["displayName"] = "MeterBarChartOverlayWithTooltip";
