// @dynatrace/strato-components/esm/charts/core/components/legend/categorical/components/ChartLegendItemLabel/ChartLegendItemLabel.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import { useRef } from "react";
import { Tooltip } from "../../../../../../../overlays/tooltip/Tooltip.js";
import { DEFAULT_TRUNCATION_MODE } from "../../../../../constants.js";
import { useChartLayoutRatio } from "../../../../chart-layout/hooks/useChartLayoutRatio.js";
import { charsFontWidth12pixelsSize500Map } from "../../../../text-ellipsis/chars-font-width-map.js";
import { ChartTextEllipsis } from "../../../../text-ellipsis/ChartTextEllipsis.js";
import { calculateWidth } from "../../../../text-ellipsis/truncate-string.js";
import { LABEL_ITEM_CLASSNAME } from "../../constants.js";
import { useLegendGridSettings } from "../../context/LegendGridSettingsContext.js";
import { useLegend } from "../../hooks/useLegend.js";
import {
  baseLegendButton,
  legendNameButton,
  legendNameButtonOverride
} from "../ChartLegendItem/ChartLegendItem.sty.js";
const MAX_WIDTH_GAP = 15;
const ChartLegendItemLabel = (props) => {
  const {
    item,
    buttonRef,
    visible,
    onButtonKeydown,
    onButtonClick,
    onMouseEnter,
    onMouseLeave
  } = props;
  const { name } = item;
  const textRef = useRef(null);
  const { readyForAutoRatio, ratio } = useChartLayoutRatio();
  const { interactionsDisabled, truncationMode } = useLegend();
  const { maxItemWidth } = useLegendGridSettings();
  let applyTruncation = false;
  if (buttonRef.current) {
    const textWidth = calculateWidth(
      item.name,
      charsFontWidth12pixelsSize500Map
    );
    const availableWidth = buttonRef.current.clientWidth;
    applyTruncation = textWidth >= availableWidth;
  }
  return /* @__PURE__ */ jsx(Tooltip, { text: name, disabled: !applyTruncation, children: /* @__PURE__ */ jsx(
    "button",
    {
      "data-testid": "chart-legend-item-name",
      "data-dt-children-name": "LegendIsolateItem",
      ref: buttonRef,
      onClick: (evt) => onButtonClick(evt.nativeEvent),
      onKeyDown: (evt) => onButtonKeydown(evt.nativeEvent),
      onMouseEnter,
      onMouseLeave,
      tabIndex: interactionsDisabled ? -1 : 0,
      className: clsx(
        baseLegendButton,
        legendNameButtonOverride,
        legendNameButton({
          isInteractive: !interactionsDisabled,
          isSubdued: !visible
        })
      ),
      style: {
        flexBasis: maxItemWidth + MAX_WIDTH_GAP
      },
      children: ratio !== "auto" || readyForAutoRatio ? /* @__PURE__ */ jsx(
        ChartTextEllipsis,
        {
          truncationMode: truncationMode ?? DEFAULT_TRUNCATION_MODE,
          "aria-label": "Series name",
          ref: textRef,
          charsFontWidthMap: charsFontWidth12pixelsSize500Map,
          children: item.name
        }
      ) : /* @__PURE__ */ jsx("span", { className: LABEL_ITEM_CLASSNAME, children: item.name })
    }
  ) });
};
export {
  ChartLegendItemLabel
};
