// @dynatrace/strato-components/charts/core/components/shared-chart-interactions/components/SynchronizedSelectionArea.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SynchronizedSelectionArea_exports = {};
__export(SynchronizedSelectionArea_exports, {
  SynchronizedSelectionArea: () => SynchronizedSelectionArea
});
module.exports = __toCommonJS(SynchronizedSelectionArea_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_useGridPlotSize = require("../../chart-grid/hooks/useGridPlotSize.js");
var import_SelectionBackdrop = require("../../selection-area/SelectionBackdrop.js");
var import_useSynchronizedSelection = require("../hooks/useSynchronizedSelection.js");
var import_SharedChartInteractions_css = require("../SharedChartInteractions.sty.js");
const SynchronizedSelectionArea = ({
  scale
}) => {
  const synchronizedSelectionArea = (0, import_useSynchronizedSelection.useSynchronizedSelection)();
  const { height } = (0, import_useGridPlotSize.useGridPlotSize)();
  const { startX, endX, resetSelectionState } = synchronizedSelectionArea;
  if (!synchronizedSelectionArea || (0, import_lodash_es.isUndefined)(startX) || (0, import_lodash_es.isUndefined)(endX)) {
    return null;
  }
  const scaledStartX = scale(startX);
  const scaledDndX = scale(endX);
  const handleOnPointerDown = (event) => {
    event.preventDefault();
    resetSelectionState?.();
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("svg", { className: import_SharedChartInteractions_css.containerSVG, onPointerDown: handleOnPointerDown, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_SelectionBackdrop.SelectionBackdrop,
    {
      focusable: false,
      interactable: false,
      startX: scaledStartX,
      endX: scaledDndX,
      height,
      isSynchronizedSelection: true
    }
  ) });
};
SynchronizedSelectionArea["displayName"] = "SynchronizedSelectionArea";
