// @dynatrace/strato-components/charts/core/components/threshold-bar/ThresholdBarTemplate.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ThresholdBarTemplate_exports = {};
__export(ThresholdBarTemplate_exports, {
  ThresholdBarTemplate: () => ThresholdBarTemplate
});
module.exports = __toCommonJS(ThresholdBarTemplate_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_intl = require("react-intl");
var import_useThresholdBarTooltip = require("./hooks/useThresholdBarTooltip.js");
var import_CircleMarker = require("./markers/CircleMarker.js");
var import_constants = require("./markers/constants.js");
var import_PillMarker = require("./markers/PillMarker.js");
var import_styles_css = require("./styles.sty.js");
var import_constants2 = require("./types/constants.js");
var import_get_threshold_int_value = require("./utils/get-threshold-int-value.js");
var import_resolve_infinite_pixel = require("./utils/resolve-infinite-pixel.js");
var import_sort_circle_markers = require("./utils/sort-circle-markers.js");
var import_OverlayTooltip = require("../overlay-tooltip/OverlayTooltip.js");
const ThresholdBarTemplate = (props) => {
  const {
    borderArea,
    length,
    markers,
    layout = "vertical",
    isFlipped = false,
    onHover
  } = props;
  const sortedMarkers = markers.toSorted(import_sort_circle_markers.sortMarkers);
  const intl = (0, import_react_intl.useIntl)();
  const {
    showTooltip,
    tooltipPosition,
    tooltipItems,
    handleMouseMove,
    handleMouseLeave
  } = (0, import_useThresholdBarTooltip.useThresholdBarTooltip)(sortedMarkers, onHover);
  const flipType = isFlipped ? "flip" : "no-flip";
  const layoutType = `${layout}-${flipType}`;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    "div",
    {
      className: (0, import_styles_css.wrapper)({
        borderArea,
        layoutType
      }),
      style: { height: length },
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)("svg", { width: import_constants2.THRESHOLD_BAR_THICKNESS, height: length, children: sortedMarkers.map((marker, index) => {
          const { y0, y1, color, isShadow = false } = marker;
          const isMarkerARange = y0 !== y1;
          const resolvedY0 = (0, import_resolve_infinite_pixel.resolveInfinitePixel)(y0, length);
          const resolvedY1 = (0, import_resolve_infinite_pixel.resolveInfinitePixel)(y1, length);
          if (isMarkerARange) {
            return [
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_PillMarker.PillMarker,
                {
                  x: 4,
                  y0: resolvedY0,
                  y1: resolvedY1,
                  color,
                  opacity: isShadow ? import_constants.MARKER_SHADOW_OPACITY : import_constants.MARKER_OPACITY,
                  onMouseMove: (event) => handleMouseMove(event, index),
                  onMouseLeave: handleMouseLeave
                },
                `threshold-bar-pill-marker-${marker.id}`
              )
            ];
          } else {
            return [
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_CircleMarker.CircleMarker,
                {
                  x: 8,
                  y: resolvedY0,
                  color,
                  onMouseMove: (event) => handleMouseMove(event, index),
                  onMouseLeave: handleMouseLeave
                },
                `threshold-bar-circle-marker-${marker.id}`
              )
            ];
          }
        }) }),
        showTooltip && tooltipPosition && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_OverlayTooltip.OverlayTooltip,
          {
            isInteractive: true,
            absoluteX: tooltipPosition.x,
            absoluteY: tooltipPosition.y,
            children: tooltipItems.map(({ id, name, color, values }) => {
              return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_OverlayTooltip.OverlayTooltip.SeriesItem,
                {
                  color: color.Default,
                  name,
                  value: `${(0, import_get_threshold_int_value.getThresholdIntValue)(intl, values)}`
                },
                id
              );
            })
          }
        )
      ]
    }
  );
};
ThresholdBarTemplate["displayName"] = "ThresholdBarTemplate";
