// @dynatrace/strato-components/forms/calendar/components/CalendarItem.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CalendarItem_exports = {};
__export(CalendarItem_exports, {
  CalendarItem: () => CalendarItem
});
module.exports = __toCommonJS(CalendarItem_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var itemStyles = __toESM(require("./CalendarItem.sty.js"));
var import_Button = require("../../../buttons/button/Button.js");
var import_useFormFieldContext = require("../../forms-core/hooks/useFormFieldContext.js");
var import_useCalendarContext = require("../hooks/useCalendarContext.js");
const CalendarItem = /* @__PURE__ */ (0, import_react.forwardRef)(
  ({
    children,
    isCurrent = false,
    highlightColor = "neutral",
    onClick,
    isSelected = false,
    isDisabled = false,
    isCurrentMonthDay = true,
    view,
    ariaLabel,
    ...remainingProps
  }, ref) => {
    const { disabled, ariaDisabled } = (0, import_useFormFieldContext.useFormFieldContext)();
    const { readOnly } = (0, import_useCalendarContext.useCalendarContext)();
    const variant = (0, import_react.useMemo)(() => {
      if (view === "day") {
        return "default";
      }
      if (isSelected) {
        return "accent";
      }
      if (isCurrent) {
        return "emphasized";
      }
      return "default";
    }, [view, isSelected, isCurrent]);
    const currentStyle = view !== "day" && isCurrent ? itemStyles.currentIndicator : "";
    const selectionHandler = !isDisabled && !disabled && !readOnly ? {
      onClick,
      onKeyDown: (event) => {
        const { code } = event;
        if (code === "Space" || code === "Enter") {
          event.preventDefault();
          onClick?.();
        }
      }
    } : {};
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Button.Button,
      {
        as: "div",
        ref,
        "aria-label": ariaLabel,
        ...remainingProps,
        className: (0, import_clsx.default)(itemStyles.calendarButton, currentStyle, {
          [itemStyles.outOfScopeDay]: !isCurrentMonthDay && highlightColor === "neutral" && !(disabled || isDisabled || ariaDisabled)
        }),
        color: highlightColor,
        disabled: disabled || isDisabled,
        "aria-disabled": ariaDisabled,
        variant,
        ...selectionHandler,
        children
      }
    );
  }
);
CalendarItem.displayName = "CalendarBodyDateButton";
