// @dynatrace/strato-components/forms/calendar/components/Day.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Day_exports = {};
__export(Day_exports, {
  Day: () => Day
});
module.exports = __toCommonJS(Day_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_date_fns = require("date-fns");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_units = require("@dynatrace-sdk/units");
var import_useTimeframeStore = require("../../../core/time/timeframeStore/useTimeframeStore.js");
var import_validation = require("../../../core/time/validation.js");
var import_useFormFieldContext = require("../../forms-core/hooks/useFormFieldContext.js");
var import_useFormFieldMessagesContext = require("../../forms-core/hooks/useFormFieldMessagesContext.js");
var styles = __toESM(require("../Calendar.sty.js"));
var import_CalendarItem = require("./CalendarItem.js");
var import_CalendarOutlineIcon = require("./CalendarOutlineIcon.js");
var import_helpers = require("../../../core/time/helpers.js");
var import_DayViewContext = require("../contexts/DayViewContext.js");
var import_useCalendarContext = require("../hooks/useCalendarContext.js");
var import_useFocus = require("../hooks/useFocus.js");
var import_calendarConfiguration = require("../types/calendarConfiguration.js");
var import_helpers2 = require("../utils/helpers.js");
const Day = ({ date, isCurrentMonthDay = true, position }) => {
  const itemRef = (0, import_react.useRef)(null);
  const isLastInRow = position % import_calendarConfiguration.NUMBER_OF_WEEKDAYS === 0;
  const isFirstInRow = position % import_calendarConfiguration.NUMBER_OF_WEEKDAYS === 1;
  const intl = (0, import_react_intl.useIntl)();
  const min = (0, import_useTimeframeStore.useCalendarStore)((store) => store.min);
  const max = (0, import_useTimeframeStore.useCalendarStore)((store) => store.max);
  const value = (0, import_useTimeframeStore.useCalendarStore)((store) => store.calendarValue ?? null);
  const setCalendarValue = (0, import_useTimeframeStore.useCalendarStore)((store) => store.setCalendarValue);
  const { state, dispatch } = (0, import_useCalendarContext.useCalendarContext)();
  const { hoveredDate, setHoveredDate } = (0, import_react.useContext)(import_DayViewContext.DayViewContext);
  const ariaLabel = (0, import_react.useMemo)(() => {
    const formatMessage = intl.formatMessage({
      defaultMessage: "Select day",
      id: "UBJSASimMVkN5w7p",
      description: "Aria label to add to each day button"
    });
    return `${formatMessage} ${(0, import_units.formatDate)(date, {
      year: "numeric",
      month: "numeric",
      day: "numeric"
    })}`;
  }, [date, intl]);
  const isCurrent = (0, import_date_fns.isSameDay)(date, state.today);
  const isFocused = (0, import_date_fns.isSameDay)(date, state.activeDate);
  const isDisabled = (0, import_react.useMemo)(() => {
    return !!((0, import_validation.isOutOfBounds)((0, import_date_fns.endOfDay)(date), min) || (0, import_validation.isOutOfBounds)(date, void 0, max));
  }, [date, min, max]);
  const { tabIndex } = (0, import_useFocus.useFocus)({ isFocused, itemRef });
  function onClick() {
    dispatch({ type: "shouldUpdateFocusFromKeydown", payload: false });
    setCalendarValue?.(date);
  }
  const {
    highlightType,
    highlightColor,
    isSelected,
    selectionVariant,
    isIncompleteSelection
  } = (0, import_react.useMemo)(
    () => (0, import_helpers2.getHighlightingProps)(date, value, state.type, hoveredDate),
    [date, value, state.type, hoveredDate]
  );
  const rangeInteractionProps = (0, import_react.useMemo)(() => {
    const onMouseLeave = ((0, import_helpers.isFullTimeframe)(value) || value === null) && hoveredDate ? () => setHoveredDate?.(null) : void 0;
    const onMouseEnter = (0, import_helpers.isPartialTimeframe)(value) ? () => setHoveredDate?.(date) : void 0;
    return {
      onMouseLeave,
      onMouseEnter
    };
  }, [value, hoveredDate, date, setHoveredDate]);
  const { disabled, ariaDisabled } = (0, import_useFormFieldContext.useFormFieldContext)();
  const {
    state: { showError }
  } = (0, import_useFormFieldMessagesContext.useFormFieldMessagesContext)();
  const validatedHighlightColor = showError && highlightColor === "primary" ? "critical" : highlightColor;
  const className = (0, import_clsx.default)(
    styles.field({
      color: validatedHighlightColor,
      isDisabled: isDisabled || disabled || ariaDisabled,
      isIncompleteSelection
    }),
    highlightType && highlightType !== "same" ? styles.rangeField({
      rangeKind: highlightType
    }) : "",
    isSelected ? styles.clearBackground : ""
  );
  const dayValue = date.getDate();
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("td", { role: "presentation", className, ...rangeInteractionProps, children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_CalendarItem.CalendarItem,
    {
      ref: itemRef,
      className: "calendar-button calendar-body-button",
      isDisabled: isDisabled || disabled,
      isCurrent,
      isSelected,
      highlightColor: validatedHighlightColor,
      isCurrentMonthDay,
      tabIndex,
      ariaLabel,
      role: "gridcell",
      "aria-selected": isSelected,
      onClick,
      view: "day",
      "data-dt-name": "SelectDay",
      "data-dt-properties": `value:${dayValue}`,
      children: [
        dayValue,
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_CalendarOutlineIcon.CalendarOutlineIcon,
          {
            isCurrent,
            variant: selectionVariant,
            highlightType: isFirstInRow && highlightType === "last" || isLastInRow && highlightType === "first" ? void 0 : highlightType
          }
        )
      ]
    }
  ) });
};
Day["displayName"] = "Day";
