// @dynatrace/strato-components/forms/calendar/components/Days.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Days_exports = {};
__export(Days_exports, {
  Days: () => Days
});
module.exports = __toCommonJS(Days_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_Day = require("./Day.js");
var styles = __toESM(require("../Calendar.sty.js"));
var import_DayViewContext = require("../contexts/DayViewContext.js");
var import_useCalendarContext = require("../hooks/useCalendarContext.js");
var import_calendarConfiguration = require("../types/calendarConfiguration.js");
var import_helpers = require("../utils/helpers.js");
function Days({ weekStartsOn }) {
  const { state } = (0, import_useCalendarContext.useCalendarContext)();
  const { setHoveredDate } = (0, import_react.useContext)(import_DayViewContext.DayViewContext);
  const days = (0, import_react.useMemo)(
    () => (0, import_helpers.getDays)(state.activeDate, weekStartsOn),
    [state.activeDate, weekStartsOn]
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "tbody",
    {
      onMouseLeave: () => {
        setHoveredDate?.(null);
      },
      children: new Array(import_calendarConfiguration.NUMBER_OF_WEEK_ROWS).fill(0).map((_, index) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        "tr",
        {
          className: (0, import_clsx.default)([styles.row, styles.dayRow]),
          children: days.slice(
            index * import_calendarConfiguration.NUMBER_OF_WEEKDAYS,
            index * import_calendarConfiguration.NUMBER_OF_WEEKDAYS + import_calendarConfiguration.NUMBER_OF_WEEKDAYS
          ).map((day, index2) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_Day.Day,
            {
              date: day.date,
              position: index2 + 1,
              isCurrentMonthDay: day.isCurrentMonthDay
            },
            day.date.toISOString()
          ))
        },
        `${days[index].date}-row-${index}`
      ))
    }
  );
}
