// @dynatrace/strato-components/forms/calendar/components/Weekdays.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Weekdays_exports = {};
__export(Weekdays_exports, {
  Weekdays: () => Weekdays
});
module.exports = __toCommonJS(Weekdays_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_tz = require("@date-fns/tz");
var import_clsx = __toESM(require("clsx"));
var import_date_fns = require("date-fns");
var import_units = require("@dynatrace-sdk/units");
var import_Text = require("../../../typography/text/Text.js");
var import_helpers = require("../../forms-core/time/helpers.js");
var styles = __toESM(require("../Calendar.sty.js"));
function Weekdays({
  weekStartsOn,
  configuredTimezone
}) {
  const today = import_tz.TZDateMini.tz(configuredTimezone);
  const intervalStart = (0, import_date_fns.startOfWeek)(today, {
    weekStartsOn
  });
  const days = (0, import_helpers.eachDayOfInterval)({
    start: intervalStart,
    end: (0, import_date_fns.addDays)(intervalStart, 6)
  });
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("thead", { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("tr", { className: (0, import_clsx.default)([styles.row, styles.dayRow, styles.weekdays]), children: days.map((weekday, index) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "th",
    {
      className: "weekday",
      abbr: (0, import_units.formatDate)(weekday, {
        weekday: "long",
        timeZone: configuredTimezone
      }),
      scope: "col",
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { textStyle: "small", as: "p", children: (0, import_units.formatDate)(weekday, {
        weekday: "short",
        timeZone: configuredTimezone
      }) })
    },
    `${weekday}-${index}`
  )) }) });
}
