// @dynatrace/strato-components/forms/calendar/components/Year.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Year_exports = {};
__export(Year_exports, {
  Year: () => Year
});
module.exports = __toCommonJS(Year_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_units = require("@dynatrace-sdk/units");
var import_Calendar_css = require("../Calendar.sty.js");
var import_CalendarItem = require("./CalendarItem.js");
var import_useFormFieldContext = require("../../forms-core/hooks/useFormFieldContext.js");
var import_useFormFieldMessagesContext = require("../../forms-core/hooks/useFormFieldMessagesContext.js");
var import_useYear = require("../hooks/useYear.js");
const Year = ({ date }) => {
  const itemRef = (0, import_react.useRef)(null);
  const { ariaLabel, tabIndex, isCurrent, isSelected, onClick, isDisabled } = (0, import_useYear.useYear)({
    date,
    itemRef
  });
  const { disabled, ariaDisabled } = (0, import_useFormFieldContext.useFormFieldContext)();
  const {
    state: { showError }
  } = (0, import_useFormFieldMessagesContext.useFormFieldMessagesContext)();
  const highlightColor = isSelected ? "primary" : "neutral";
  const validatedHighlightColor = showError && highlightColor === "primary" ? "critical" : highlightColor;
  const yearValue = (0, import_units.formatDate)(date, { year: "numeric" });
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "td",
    {
      role: "presentation",
      className: (0, import_Calendar_css.field)({
        color: validatedHighlightColor,
        isDisabled: isDisabled || disabled || ariaDisabled
      }),
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_CalendarItem.CalendarItem,
        {
          ref: itemRef,
          className: "calendar-button calendar-body-button",
          isCurrent,
          isSelected,
          isDisabled: isDisabled || disabled,
          highlightColor: validatedHighlightColor,
          tabIndex,
          role: "gridcell",
          "aria-selected": !!isSelected,
          ariaLabel,
          onClick,
          view: "year",
          "data-dt-name": "SelectYear",
          "data-dt-properties": `value:${yearValue}`,
          children: yearValue
        }
      )
    }
  );
};
Year["displayName"] = "Year";
