// @dynatrace/strato-components/forms/forms-core/hooks/useFormFieldValidation.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useFormFieldValidation_exports = {};
__export(useFormFieldValidation_exports, {
  getCombinedAriaProps: () => getCombinedAriaProps,
  getErrorType: () => getErrorType,
  getI18nErrorMessage: () => getI18nErrorMessage,
  useFormFieldValidation: () => useFormFieldValidation,
  useFormFieldValidationAriaProps: () => useFormFieldValidationAriaProps
});
module.exports = __toCommonJS(useFormFieldValidation_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_useFormFieldMessagesContext = require("./useFormFieldMessagesContext.js");
var import_useId = require("../../../core/hooks/useId.js");
var import_Text = require("../../../typography/text/Text.js");
var import_formfield_messages_state_reducer = require("../state/formfield-messages-state-reducer.js");
function getErrorType(validityState) {
  if (!validityState) {
    return null;
  }
  if (validityState.badInput) {
    return "badInput";
  }
  if (validityState.patternMismatch) {
    return "patternMismatch";
  }
  if (validityState.rangeOverflow) {
    return "rangeOverflow";
  }
  if (validityState.rangeUnderflow) {
    return "rangeUnderflow";
  }
  if (validityState.stepMismatch) {
    return "stepMismatch";
  }
  if (validityState.tooLong) {
    return "tooLong";
  }
  if (validityState.tooShort) {
    return "tooShort";
  }
  if (validityState.typeMismatch) {
    return "typeMismatch";
  }
  if (validityState.valueMissing) {
    return "valueMissing";
  }
  return null;
}
function getI18nErrorMessage(key, ref) {
  switch (key) {
    case "badInput":
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_react_intl.FormattedMessage,
        {
          id: "/cwiOkubC69MZmRu",
          defaultMessage: "Please provide input in a valid format.",
          description: "Error message for input that the browser is unable to convert, e.g. string for a number input."
        }
      );
    case "patternMismatch":
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_react_intl.FormattedMessage,
        {
          id: "qPeX7AgKbKzBFqGZ",
          defaultMessage: "Please match the format requested.",
          description: "Error message for input that doesn't follow the requested pattern."
        }
      );
    case "rangeOverflow":
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_react_intl.FormattedMessage,
        {
          id: "14mVeqEFpqD2fpev",
          defaultMessage: "Value must be less or equal to {max}.",
          description: "Error message for input that value is higher then the max.",
          values: { max: ref?.getAttribute("max") }
        }
      );
    case "rangeUnderflow":
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_react_intl.FormattedMessage,
        {
          id: "5XoVjJUQ5b86MLW+",
          defaultMessage: "Value must be higher or equal to {min}.",
          description: "Error message for input that value is lower then the max.",
          values: { min: ref?.getAttribute("min") }
        }
      );
    case "stepMismatch":
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_react_intl.FormattedMessage,
        {
          id: "MYfVMxPsAgOn8JSP",
          defaultMessage: "Value must be a multiple of {step}, starting from {min}.",
          description: "Only values which are a whole number of steps from the step base are valid. The step base is min if specified, value otherwise, or 0 if neither is provided.",
          values: {
            step: ref?.getAttribute("step"),
            min: ref?.getAttribute("min") ?? "0"
          }
        }
      );
    case "tooLong":
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_react_intl.FormattedMessage,
        {
          id: "JOIA3jQfIIkgnW04",
          defaultMessage: "Please shorten this text to {maxLength} characters or less.",
          description: "Error message for input that is shorter than the maxLength.",
          values: { maxLength: ref?.getAttribute("maxLength") }
        }
      );
    case "tooShort":
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_react_intl.FormattedMessage,
        {
          id: "sJeuQTGrzcMvd6fY",
          defaultMessage: "Please lengthen this text to {minLength} characters or more.",
          description: "Error message for input that is shorter than the minLength.",
          values: { minLength: ref?.getAttribute("minLength") }
        }
      );
    case "typeMismatch":
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_react_intl.FormattedMessage,
        {
          id: "fLfhKO8cXlo3vW5M",
          defaultMessage: "Please provide a value that matches the requested type ({type}).",
          description: "Error message for input that doesn't match the expected type.",
          values: { type: ref?.getAttribute("type") }
        }
      );
    case "valueMissing":
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_react_intl.FormattedMessage,
        {
          id: "laIylhVFbIvEZrxt",
          defaultMessage: "Please fill in this field.",
          description: "Error message for missing input."
        }
      );
    case "customError":
    case "valid":
    default:
      return null;
  }
}
function useFormFieldValidationAriaProps(ariaLabelingProps = {}) {
  const { state } = (0, import_useFormFieldMessagesContext.useFormFieldMessagesContext)();
  const _ariaProps = (0, import_react.useMemo)(() => {
    let messageIdsString;
    let errorMessageIdsString;
    if (state.customMessages.length > 0 || state.formControlMessages.length > 0) {
      const messages = [...state.customMessages, ...state.formControlMessages];
      errorMessageIdsString = messages.filter((msg) => msg.variant === "error").map((msg) => msg.id).join(" ");
      messageIdsString = messages.map((msg) => msg.id).join(" ");
    }
    return {
      "aria-errormessage": state.showError ? errorMessageIdsString : void 0,
      "aria-invalid": state.showError,
      "aria-describedby": messageIdsString
    };
  }, [state.customMessages, state.formControlMessages, state.showError]);
  const ariaProps = getCombinedAriaProps(_ariaProps, ariaLabelingProps);
  return ariaProps;
}
function useFormFieldValidation(ref, validationDeps = {}, ariaLabelingProps = {}, controlId) {
  const { state, dispatch } = (0, import_useFormFieldMessagesContext.useFormFieldMessagesContext)();
  const errorId = (0, import_useId.useId)();
  const ariaProps = useFormFieldValidationAriaProps(ariaLabelingProps);
  const formControlMessagesRef = (0, import_react.useRef)(state.formControlMessages);
  formControlMessagesRef.current = state.formControlMessages;
  const updateErrorMessages = (0, import_react.useCallback)(() => {
    const errorType = getErrorType(ref.current?.validity);
    formControlMessagesRef.current.forEach((message) => {
      if (message.validityStateKey !== errorType && message.validityStateKey !== "customError") {
        dispatch?.({
          type: import_formfield_messages_state_reducer.FormFieldMessagesAction.REMOVEMESSAGE,
          payload: {
            id: message.id,
            messageType: "formcontrol",
            controlId
          }
        });
      }
    });
    if (errorType) {
      dispatch?.({
        type: import_formfield_messages_state_reducer.FormFieldMessagesAction.ADDMESSAGE,
        payload: {
          id: errorId,
          message: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { textStyle: "small", as: "p", children: getI18nErrorMessage(errorType, ref.current) }),
          validityStateKey: errorType,
          messageType: "formcontrol",
          variant: "error",
          controlId
        }
      });
    }
  }, [controlId, dispatch, errorId, ref]);
  const onInvalid = (0, import_react.useCallback)(
    (event) => {
      updateErrorMessages();
      dispatch?.({
        type: import_formfield_messages_state_reducer.FormFieldMessagesAction.SETISTOUCHED,
        payload: { isTouched: true }
      });
      event.preventDefault();
    },
    [dispatch, updateErrorMessages]
  );
  const onBlur = (0, import_react.useCallback)(() => {
    ref.current?.reportValidity();
    dispatch?.({
      type: import_formfield_messages_state_reducer.FormFieldMessagesAction.SETISTOUCHED,
      payload: { isTouched: true }
    });
  }, [dispatch, ref]);
  const validationDepsKey = JSON.stringify(validationDeps);
  (0, import_react.useEffect)(() => {
    updateErrorMessages();
  }, [validationDepsKey, updateErrorMessages]);
  return {
    ariaProps,
    onInvalid,
    onChange: updateErrorMessages,
    onBlur
  };
}
function getCombinedAriaProps(validationAriaProps, labelingProps) {
  const {
    "aria-describedby": validationDescription,
    ...remainingValidationProps
  } = validationAriaProps;
  const { "aria-describedby": propDescription, ...remainingLabelingProps } = labelingProps;
  if (validationDescription && propDescription) {
    return {
      ...remainingLabelingProps,
      ...remainingValidationProps,
      "aria-describedby": (validationDescription + " " + propDescription).trim()
    };
  }
  return {
    ...remainingLabelingProps,
    ...remainingValidationProps,
    "aria-describedby": validationDescription ?? propDescription
  };
}
