// @dynatrace/strato-components/forms/forms-core/hooks/useSyncedFormControlProps.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useSyncedFormControlProps_exports = {};
__export(useSyncedFormControlProps_exports, {
  useSyncedFormControlProps: () => useSyncedFormControlProps
});
module.exports = __toCommonJS(useSyncedFormControlProps_exports);
var import_react = require("react");
var import_useFormFieldContext = require("./useFormFieldContext.js");
var import_useFormFieldMessagesContext = require("./useFormFieldMessagesContext.js");
var import_useId = require("../../../core/hooks/useId.js");
var import_parse_boolean = require("../../../core/utils/parse-boolean.js");
var import_formfield_messages_state_reducer = require("../state/formfield-messages-state-reducer.js");
var import_formfield_state_reducer = require("../state/formfield-state-reducer.js");
function useSyncedFormControlProps(formcontrolprops) {
  const uniqueId = (0, import_useId.useId)();
  const { propDisabled, propRequired, propId, propAriaDisabled, propReadOnly } = formcontrolprops;
  const {
    controlId,
    required: contextRequired,
    disabled: contextDisabled,
    ariaDisabled: contextAriaDisabled,
    readOnly: contextReadOnly,
    dispatch
  } = (0, import_useFormFieldContext.useFormFieldContext)();
  const { dispatch: formFieldMessageContextDispatch } = (0, import_useFormFieldMessagesContext.useFormFieldMessagesContext)();
  const disabled = propDisabled ?? contextDisabled;
  const ariaDisabled = (0, import_parse_boolean.parseBoolean)(propAriaDisabled) ?? contextAriaDisabled;
  const required = propRequired ?? contextRequired;
  const readOnly = propReadOnly ?? contextReadOnly;
  const id = propId || controlId || uniqueId;
  (0, import_react.useEffect)(() => {
    dispatch?.({
      type: import_formfield_state_reducer.FormFieldAction.UPDATE,
      payload: {
        id: propId,
        disabled: propDisabled,
        required: propRequired,
        ariaDisabled: (0, import_parse_boolean.parseBoolean)(propAriaDisabled),
        readOnly: propReadOnly
      }
    });
  }, [
    dispatch,
    propDisabled,
    propId,
    propRequired,
    propAriaDisabled,
    propReadOnly
  ]);
  (0, import_react.useEffect)(() => {
    formFieldMessageContextDispatch?.({
      type: import_formfield_messages_state_reducer.FormFieldMessagesAction.DISABLEMESSAGE,
      payload: { isDisabled: disabled }
    });
  }, [disabled, formFieldMessageContextDispatch]);
  return {
    disabled,
    required,
    ariaDisabled,
    id,
    readOnly
  };
}
