// @dynatrace/strato-geo/map/shapes/shape-picker.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var shape_picker_exports = {};
__export(shape_picker_exports, {
  getShape: () => getShape
});
module.exports = __toCommonJS(shape_picker_exports);
var import_circle = require("./circle.js");
var import_cross = require("./cross.js");
var import_diamond = require("./diamond.js");
var import_heart = require("./heart.js");
var import_pin = require("./pin.js");
var import_square = require("./square.js");
var import_star = require("./star.js");
var import_triangle = require("./triangle.js");
const getShape = (shape) => {
  const shapePicker = {
    pin: import_pin.pin,
    circle: import_circle.circle,
    square: import_square.square,
    diamond: import_diamond.diamond,
    star: import_star.star,
    heart: import_heart.heart,
    cross: import_cross.cross,
    triangle: import_triangle.triangle
  };
  return shapePicker[shape];
};
