// @dynatrace/strato-components/esm/content/markdown/Markdown.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import { forwardRef, useEffect, useMemo, useRef } from "react";
import ReactMarkdown from "react-markdown";
import gfm from "remark-gfm";
import { MarkdownWrapperCSS } from "./Markdown.sty.js";
import { tableMappings } from "./transformers/table-mapping.js";
import { typographyMappings } from "./transformers/typography-mapping.js";
import { findValidParent, findOffset, extractSourcePos } from "./utils.js";
import { useAriaLabelingProps } from "../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useSafeForwardProps } from "../../core/hooks/useSafeForwardProps.js";
import { textStyle as textStyleCSS } from "../../styles/textStyle.sty.js";
const COMPONENT_NAME = "Markdown";
const componentMappings = {
  ...typographyMappings,
  ...tableMappings
};
const Markdown = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const {
    children,
    onSelectionChange,
    customComponentMappings,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
  const ariaLabelingProps = useAriaLabelingProps(restProps);
  const forwardProps = useSafeForwardProps(restProps, ariaLabelingProps);
  const startNode = useRef(null);
  const endNode = useRef(null);
  const extendedComponentMappings = useMemo(() => {
    return customComponentMappings ? {
      ...componentMappings,
      ...customComponentMappings
    } : componentMappings;
  }, [customComponentMappings]);
  const handleSelectionChange = () => {
    const selection = document.getSelection();
    if (!selection || selection.isCollapsed || !selection.anchorNode || !selection.focusNode) {
      return;
    }
    const start = findValidParent(selection.anchorNode);
    const end = findValidParent(selection.focusNode);
    if (!start && !end) {
      return;
    }
    if (start) {
      startNode.current = start;
    }
    if (end) {
      endNode.current = end;
    }
    if (startNode.current && endNode.current) {
      const sourcePos = extractSourcePos(startNode.current, endNode.current);
      if (sourcePos && onSelectionChange) {
        const startOffset = findOffset(sourcePos.start, children, -1);
        const endOffset = findOffset(sourcePos?.end, children);
        sourcePos.start.offset = startOffset;
        sourcePos.end.offset = endOffset;
        onSelectionChange(sourcePos);
      }
    }
  };
  useEffect(() => {
    document.addEventListener("selectionchange", handleSelectionChange);
    return () => {
      document.removeEventListener("selectionchange", handleSelectionChange);
    };
  }, [children]);
  return /* @__PURE__ */ jsx(
    "div",
    {
      ref: forwardedRef,
      className: clsx(
        textStyleCSS({ textStyle: "base" }),
        MarkdownWrapperCSS,
        consumerClassName
      ),
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...ariaLabelingProps,
      ...behavioralTrackingProps,
      ...forwardProps,
      "data-dt-component": COMPONENT_NAME,
      children: /* @__PURE__ */ jsx(
        ReactMarkdown,
        {
          components: extendedComponentMappings,
          remarkPlugins: [gfm],
          skipHtml: true,
          sourcePos: true,
          children
        }
      )
    }
  );
});
Markdown.displayName = COMPONENT_NAME;
export {
  Markdown,
  componentMappings
};
