// @dynatrace/strato-components/esm/content/markdown/utils.js
import { isNaN, isNil } from "lodash-es";
import { Children, isValidElement } from "react";
function hasChildren(element) {
  if (isValidElement(element)) {
    return !!element?.props?.children;
  }
  return false;
}
function getChildren(element) {
  return Children.toArray(hasChildren(element) ? element.props.children : []);
}
function matchLanguageParm(className) {
  return /language-(\w+)/.exec(className || "");
}
function extractSourcePos(start, end) {
  const regexp = /^(\d+):(\d+)-(\d+):(\d+)/;
  const match1 = regexp.exec(
    start.getAttribute("data-sourcepos") ?? ""
  );
  const match2 = regexp.exec(
    end.getAttribute("data-sourcepos") ?? ""
  );
  if (match1 && match2) {
    const [, start1, sOffset1, end1, eOffset1] = match1.filter((m) => !isNaN(m)).map((m) => Number(m));
    const [, start2, sOffset2, end2, eOffset2] = match2.filter((m) => !isNaN(m)).map((m) => Number(m));
    return {
      start: {
        line: start1 < start2 ? start1 : start2,
        column: start1 < start2 ? sOffset1 : sOffset2,
        offset: 0
      },
      end: {
        line: end1 > end2 ? end1 : end2,
        column: end1 > end2 ? eOffset1 : eOffset2,
        offset: 0
      }
    };
  }
  return null;
}
function findValidParent(element, count = 0) {
  if (count > 10 || element === null) {
    return null;
  }
  if (element instanceof Element && element.getAttribute("data-sourcepos")) {
    return element;
  }
  return findValidParent(element.parentNode, count + 1);
}
function findOffset(position, source, lineOffset = 0) {
  let offset = 0;
  const lines = source.split("\n");
  for (let i = 0; i < position.line + lineOffset; i++) {
    if (!isNil(lines[i])) {
      offset += lines[i].length + 1;
    }
  }
  return offset;
}
export {
  extractSourcePos,
  findOffset,
  findValidParent,
  getChildren,
  matchLanguageParm
};
