// @dynatrace/strato-components/esm/tables/SimpleTable/SimpleTable.js
import { jsx, jsxs } from "react/jsx-runtime";
import {
  flexRender,
  getCoreRowModel,
  useReactTable
} from "@tanstack/react-table";
import clsx from "clsx";
import { forwardRef } from "react";
import { useColumns } from "./hooks/useColumns.js";
import * as styles from "./SimpleTable.sty.js";
import { useAriaLabelingProps } from "../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useSafeForwardProps } from "../../core/hooks/useSafeForwardProps.js";
const COMPONENT_NAME = "SimpleTable";
function InternalSimpleTable(props, forwardedRef) {
  const {
    data,
    columns,
    variant,
    style: consumerStyle,
    className: consumerClassName,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
  const ariaLabelingProps = useAriaLabelingProps(restProps);
  const forwardProps = useSafeForwardProps(restProps, ariaLabelingProps);
  const extractedColumns = useColumns(columns);
  const table = useReactTable({
    data,
    columns: extractedColumns,
    getCoreRowModel: getCoreRowModel()
  });
  const combinedVariants = {
    rowSeparation: variant?.rowSeparation ?? "horizontalDividers",
    verticalDividers: variant?.verticalDividers ?? false,
    contained: variant?.contained ?? true,
    rowDensity: variant?.rowDensity ?? "default",
    fontStyle: variant?.fontStyle ?? "text"
  };
  return /* @__PURE__ */ jsxs(
    "table",
    {
      ref: forwardedRef,
      className: clsx(styles.simpleTable, consumerClassName, {
        "strato-table-contained": combinedVariants.contained,
        "strato-table-colsep-divider": combinedVariants.verticalDividers,
        "strato-table-code-font": combinedVariants.fontStyle === "code",
        "strato-table-cell-density-default": combinedVariants.rowDensity === "default",
        "strato-table-cell-density-condensed": combinedVariants.rowDensity === "condensed",
        "strato-table-cell-density-comfortable": combinedVariants.rowDensity === "comfortable",
        "strato-table-rowsep-zebra": combinedVariants.rowSeparation === "zebraStripes",
        "strato-table-rowsep-divider": combinedVariants.rowSeparation === "horizontalDividers" || combinedVariants.rowSeparation === "zebraStripes"
      }),
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...ariaLabelingProps,
      ...behavioralTrackingProps,
      ...forwardProps,
      "data-dt-component": COMPONENT_NAME,
      children: [
        /* @__PURE__ */ jsx("thead", { children: table.getHeaderGroups().map((headerGroup) => /* @__PURE__ */ jsx("tr", { className: "strato-table-header", children: headerGroup.headers.map((header) => /* @__PURE__ */ jsx(
          "th",
          {
            className: clsx("strato-table-header-cell", {
              "strato-table-cell-alignment-left": header.column.columnDef.meta?.alignment === "left",
              "strato-table-cell-alignment-center": header.column.columnDef.meta?.alignment === "center",
              "strato-table-cell-alignment-right": header.column.columnDef.meta?.alignment === "right",
              "strato-table-cell-alignment-top-left": header.column.columnDef.meta?.alignment === "top-left",
              "strato-table-cell-alignment-top-center": header.column.columnDef.meta?.alignment === "top-center",
              "strato-table-cell-alignment-top-right": header.column.columnDef.meta?.alignment === "top-right",
              "strato-table-cell-alignment-bottom-left": header.column.columnDef.meta?.alignment === "bottom-left",
              "strato-table-cell-alignment-bottom-center": header.column.columnDef.meta?.alignment === "bottom-center",
              "strato-table-cell-alignment-bottom-right": header.column.columnDef.meta?.alignment === "bottom-right"
            }),
            style: {
              minWidth: header.column.columnDef.meta?.minWidth,
              width: header.column.columnDef.meta?.width,
              maxWidth: header.column.columnDef.meta?.maxWidth
            },
            children: header.isPlaceholder ? null : flexRender(
              header.column.columnDef.header,
              header.getContext()
            )
          },
          header.id
        )) }, headerGroup.id)) }),
        /* @__PURE__ */ jsx("tbody", { children: table.getRowModel().rows.map((row, rowIndex) => /* @__PURE__ */ jsx(
          "tr",
          {
            className: clsx("strato-table-row", {
              "strato-table-row-even": rowIndex % 2
            }),
            children: row.getVisibleCells().map((cell) => /* @__PURE__ */ jsx(
              "td",
              {
                className: clsx("strato-table-cell", {
                  "strato-table-cell-alignment-left": cell.column.columnDef.meta?.alignment === "left",
                  "strato-table-cell-alignment-center": cell.column.columnDef.meta?.alignment === "center",
                  "strato-table-cell-alignment-right": cell.column.columnDef.meta?.alignment === "right",
                  "strato-table-cell-alignment-top-left": cell.column.columnDef.meta?.alignment === "top-left",
                  "strato-table-cell-alignment-top-center": cell.column.columnDef.meta?.alignment === "top-center",
                  "strato-table-cell-alignment-top-right": cell.column.columnDef.meta?.alignment === "top-right",
                  "strato-table-cell-alignment-bottom-left": cell.column.columnDef.meta?.alignment === "bottom-left",
                  "strato-table-cell-alignment-bottom-center": cell.column.columnDef.meta?.alignment === "bottom-center",
                  "strato-table-cell-alignment-bottom-right": cell.column.columnDef.meta?.alignment === "bottom-right"
                }),
                style: {
                  minWidth: cell.column.columnDef.meta?.minWidth,
                  width: cell.column.columnDef.meta?.width,
                  maxWidth: cell.column.columnDef.meta?.maxWidth
                },
                children: flexRender(cell.column.columnDef.cell, cell.getContext())
              },
              cell.id
            ))
          },
          row.id
        )) })
      ]
    }
  );
}
InternalSimpleTable.displayName = COMPONENT_NAME;
const SimpleTable = /* @__PURE__ */ forwardRef(InternalSimpleTable);
export {
  SimpleTable
};
