// @dynatrace/strato-components/esm/charts/core/components/tooltip/InteractiveChartTooltip/components/InteractiveChartTooltipVirtualizedSeries.js
import { jsx, jsxs } from "react/jsx-runtime";
import { useVirtualizer } from "@tanstack/react-virtual";
import { useEffect, useRef } from "react";
import {
  TooltipItemsContainer,
  TooltipItems,
  StyledTooltipContainer,
  ScrollableContainerParent,
  DEFAULT_TOOLTIP_MIN_WIDTH,
  TOOLTIP_MAX_HEIGHT
} from "./InteractiveChartTooltip.styles.js";
import { Flex } from "../../../../../../layouts/flex/Flex.js";
import { useHandleSetScrollbar } from "../hooks/useHandleSetScrollbar.js";
const InteractiveChartTooltipVirtualizedSeries = (props) => {
  const { itemSize, header, body, pointerEvents, onScrollContent } = props;
  const scrollRef = useRef(null);
  const scrollRefContainer = useRef(null);
  const { getVirtualItems, getTotalSize, measureElement } = useVirtualizer({
    count: body?.length ?? 0,
    getScrollElement: () => scrollRefContainer.current,
    estimateSize: (index) => {
      return itemSize?.(index) ?? 16;
    },
    overscan: 3
  });
  const items = getVirtualItems();
  useHandleSetScrollbar(getTotalSize());
  useEffect(() => {
    if (!onScrollContent) {
      return;
    }
    const scrollContainer = scrollRefContainer.current;
    scrollContainer?.addEventListener("scroll", onScrollContent);
    return () => {
      scrollContainer?.removeEventListener("scroll", onScrollContent);
    };
  }, [scrollRefContainer, onScrollContent]);
  const containerMinHeight = typeof itemSize === "undefined" ? Math.min((body?.length ?? 0) * 16, TOOLTIP_MAX_HEIGHT) : Math.min((body?.length ?? 0) * itemSize(0), TOOLTIP_MAX_HEIGHT);
  return /* @__PURE__ */ jsx(
    StyledTooltipContainer,
    {
      minWidth: DEFAULT_TOOLTIP_MIN_WIDTH,
      allowPointerEvents: pointerEvents,
      children: /* @__PURE__ */ jsxs(Flex, { gap: 0, flexDirection: "column", children: [
        header,
        /* @__PURE__ */ jsx(
          ScrollableContainerParent,
          {
            ref: scrollRefContainer,
            "data-testid": "chart-tooltip-item-container",
            style: {
              minHeight: `clamp(${containerMinHeight}px, ${getTotalSize()}px, ${TOOLTIP_MAX_HEIGHT}px)`
            },
            children: /* @__PURE__ */ jsx(
              TooltipItemsContainer,
              {
                ref: scrollRef,
                style: {
                  height: getTotalSize()
                },
                children: /* @__PURE__ */ jsx(
                  "div",
                  {
                    style: {
                      transform: `translateY(${items[0]?.start ?? 0}px)`
                    },
                    children: items.map(({ index }) => {
                      const CurrentVirtualizedItem = body[index];
                      return /* @__PURE__ */ jsx(
                        TooltipItems,
                        {
                          "data-testid": "chart-tooltip-item",
                          "data-index": index,
                          ref: measureElement,
                          children: CurrentVirtualizedItem()
                        },
                        index
                      );
                    })
                  }
                )
              }
            )
          }
        )
      ] })
    }
  );
};
InteractiveChartTooltipVirtualizedSeries["displayName"] = "InteractiveChartTooltipSeries";
export {
  InteractiveChartTooltipVirtualizedSeries
};
