// @dynatrace/strato-components/esm/charts/core/components/tooltip/InteractiveChartTooltip/components/InteractiveChartTooltip.js
import { jsx } from "react/jsx-runtime";
import {
  forwardRef,
  useEffect,
  useRef,
  useState,
  Children
} from "react";
import {
  BasicChartTooltip
} from "./BasicChartTooltip/BasicChartTooltip.js";
import { InteractiveChartTooltipVirtualizedSeries } from "./InteractiveChartTooltipVirtualizedSeries.js";
import { useMergeRefs } from "../../../../../../core/hooks/useMergeRefs.js";
import { isSlot } from "../../../../../../core/utils/is-slot.js";
import { isTouchableDevice } from "../../../../utils/is-touchable-device.js";
import { useChartTooltip } from "../../shared/hooks/useChartTooltip.js";
import { InteractiveTooltipContext } from "../context/InteractiveTooltipContext.js";
import { Actions } from "../slots/Actions.js";
import { Body } from "../slots/Body.js";
import { Header } from "../slots/Header.js";
function iterateSlots(children) {
  const slots = { Header: void 0, Body: void 0, Actions: void 0 };
  Children.forEach(children, (child) => {
    if (isSlot(child, Header)) {
      slots.Header = child.props.children;
    } else if (isSlot(child, Body)) {
      slots.Body = child.props.children;
    } else if (isSlot(child, Actions)) {
      slots.Actions = child.props.children;
    }
  });
  return slots;
}
const _InteractiveChartTooltip = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const {
    pinned = false,
    chartRef,
    position,
    itemSize,
    updatePositionWhenPinned = false,
    children,
    onMouseLeave,
    onScrollContent
  } = props;
  const chartTooltipRef = useRef(null);
  const mergedRefs = useMergeRefs([forwardedRef, chartTooltipRef]);
  const [selectedItem, selectItem] = useState(
    void 0
  );
  const [hasScrollbar, setHasScrollbar] = useState(false);
  const onItemSelected = (e, item) => {
    e.stopPropagation();
    selectItem(item);
  };
  const { placement, updateTooltipPosition } = useChartTooltip(
    chartTooltipRef.current
  );
  useEffect(() => {
    if (!pinned || isTouchableDevice() || updatePositionWhenPinned) {
      updateTooltipPosition(position);
    }
  }, [
    pinned,
    position.x,
    position.y,
    updatePositionWhenPinned,
    updateTooltipPosition
  ]);
  useEffect(() => {
    return () => {
      onMouseLeave && onMouseLeave();
    };
  }, []);
  const { Header: Header2, Body: Body2 } = iterateSlots(children);
  return /* @__PURE__ */ jsx(
    InteractiveTooltipContext.Provider,
    {
      value: {
        selectedItem,
        selectItem: onItemSelected,
        hasScrollbar,
        setHasScrollbar
      },
      children: /* @__PURE__ */ jsx(
        BasicChartTooltip,
        {
          ref: mergedRefs,
          placement,
          chartRef,
          onMouseLeave,
          pinned,
          children: /* @__PURE__ */ jsx(
            InteractiveChartTooltipVirtualizedSeries,
            {
              pinned,
              itemSize,
              header: Header2,
              body: Body2,
              pointerEvents: pinned,
              onScrollContent
            }
          )
        }
      )
    }
  );
});
_InteractiveChartTooltip.displayName = "InteractiveChartTooltip";
const InteractiveChartTooltip = Object.assign(_InteractiveChartTooltip, {
  Header,
  Body,
  Actions
});
export {
  InteractiveChartTooltip
};
