// @dynatrace/strato-components/esm/charts/core/components/tooltip/InteractiveChartTooltip/components/InteractiveChartTooltip.styles.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import { forwardRef } from "react";
import Spacings from "@dynatrace/strato-design-tokens/spacings";
import * as styles from "./InteractiveChartTooltip.sty.js";
import { Flex } from "../../../../../../layouts/flex/Flex.js";
const TOOLTIP_MAX_HEIGHT = 200;
const TOOLTIP_MAX_WIDTH = 300;
const DEFAULT_TOOLTIP_MIN_WIDTH = "225px";
function StyledChartTooltipHeader(props) {
  const { className, ...remainingProps } = props;
  return /* @__PURE__ */ jsx(
    "div",
    {
      ...remainingProps,
      className: clsx(styles.tooltipHeader, className),
      tabIndex: -1
    }
  );
}
function StyledTitleContainer(props) {
  const { className, as = "div", ...remainingProps } = props;
  return /* @__PURE__ */ jsx(
    Flex,
    {
      ...remainingProps,
      as,
      padding: 4,
      className: clsx(styles.titleContainer, className)
    }
  );
}
function StyledTooltipContainer(props) {
  const { style, minWidth, allowPointerEvents, ...remainingProps } = props;
  return /* @__PURE__ */ jsx(
    "div",
    {
      ...remainingProps,
      style: {
        ...style,
        minWidth,
        maxWidth: `calc(${TOOLTIP_MAX_WIDTH}px - (${Spacings.Size8} * 2))`,
        pointerEvents: allowPointerEvents ? "auto" : "none",
        cursor: allowPointerEvents ? "pointer" : "auto"
      }
    }
  );
}
function StyledFormattedTimeframe(props) {
  const { className, ...remainingProps } = props;
  return /* @__PURE__ */ jsx(
    "span",
    {
      ...remainingProps,
      className: clsx(styles.formattedTimeframe, className)
    }
  );
}
const TooltipItemsContainer = forwardRef((props, ref) => {
  const { className, ...remainingProps } = props;
  return /* @__PURE__ */ jsx(
    "div",
    {
      ref,
      ...remainingProps,
      className: clsx(styles.tooltipItemsContainer, className)
    }
  );
});
const ScrollableContainerParent = forwardRef((props, ref) => {
  const { className, style, ...remainingProps } = props;
  return /* @__PURE__ */ jsx(
    "div",
    {
      ref,
      ...remainingProps,
      className: clsx(styles.scrollableContainerParent, className),
      style: { ...style ?? {}, maxHeight: TOOLTIP_MAX_HEIGHT }
    }
  );
});
const TooltipItems = forwardRef(
  (props, ref) => {
    const { className, ...remainingProps } = props;
    return /* @__PURE__ */ jsx(
      "div",
      {
        ref,
        ...remainingProps,
        className: clsx(styles.tooltipItem, className)
      }
    );
  }
);
export {
  DEFAULT_TOOLTIP_MIN_WIDTH,
  ScrollableContainerParent,
  StyledChartTooltipHeader,
  StyledFormattedTimeframe,
  StyledTitleContainer,
  StyledTooltipContainer,
  TOOLTIP_MAX_HEIGHT,
  TooltipItems,
  TooltipItemsContainer
};
