// @dynatrace/strato-components/esm/charts/core/components/tooltip/InteractiveChartTooltip/components/SeriesName.js
import { jsx } from "react/jsx-runtime";
import { useRef } from "react";
import { OverlayStackingProvider } from "../../../../../../core/providers/OverlayStackingProvider.js";
import { Tooltip } from "../../../../../../overlays/tooltip/Tooltip.js";
import { DEFAULT_TRUNCATION_MODE } from "../../../../constants.js";
import { useTruncationDetector } from "../../../../hooks/useTruncationDetector.js";
import { singleLineName } from "../../../../utils/single-line-name.js";
import { ChartTextEllipsis } from "../../../text-ellipsis/ChartTextEllipsis.js";
import {
  multipleDimensionItem,
  multipleDimensionContainer,
  singleDimensionContainer
} from "../../shared/styles/SeriesBasedContent.sty.js";
const SeriesName = ({
  names,
  item,
  nameFormatter,
  singleLine,
  highlighted,
  truncationMode,
  isScrolling
}) => {
  const textRef = useRef(null);
  const name = nameFormatter?.(item) ?? singleLineName(item.name);
  const isTruncated = useTruncationDetector(textRef, name);
  return /* @__PURE__ */ jsx(OverlayStackingProvider, { children: /* @__PURE__ */ jsx(Tooltip, { text: name, disabled: isScrolling || !isTruncated, children: singleLine ? /* @__PURE__ */ jsx("div", { className: singleDimensionContainer({ highlighted }), children: /* @__PURE__ */ jsx(
    ChartTextEllipsis,
    {
      ref: textRef,
      truncationMode: truncationMode ?? DEFAULT_TRUNCATION_MODE,
      "aria-label": "Series name",
      children: name
    }
  ) }) : /* @__PURE__ */ jsx("div", { className: multipleDimensionContainer({ highlighted }), children: names.map((dim, index) => /* @__PURE__ */ jsx(
    "span",
    {
      className: multipleDimensionItem,
      children: /* @__PURE__ */ jsx(
        ChartTextEllipsis,
        {
          truncationMode: truncationMode ?? DEFAULT_TRUNCATION_MODE,
          "aria-label": "Series name",
          children: dim
        }
      )
    },
    `dimension-${index}`
  )) }) }) });
};
SeriesName["displayName"] = "SeriesName";
export {
  SeriesName
};
