// @dynatrace/strato-components/esm/tables/DataTable/components/SettingsModal/DataTableSettingsModal.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { forwardRef } from "react";
import { useIntl } from "react-intl";
import { Button } from "../../../../buttons/button/Button.js";
import { useId } from "../../../../core/hooks/useId.js";
import { Flex } from "../../../../layouts/flex/Flex.js";
import { Modal } from "../../../../overlays/modal/Modal.js";
import { screenReaderOnly } from "../../../../styles/screen-reader-only.sty.js";
import { dataTableMessages } from "../../messages.js";
const DataTableSettingsModal = forwardRef((props, forwardedRef) => {
  const {
    title,
    dismissible = true,
    show,
    showResetStatus = false,
    children,
    onDismiss,
    onReset,
    onApply,
    controlLabels,
    disableButtons,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    className: consumerClassName,
    style: consumerStyle
  } = props;
  const intl = useIntl();
  const statusId = useId();
  const showFooter = onDismiss || onReset || onApply;
  return /* @__PURE__ */ jsx(
    Modal,
    {
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ref: forwardedRef,
      className: consumerClassName,
      style: consumerStyle,
      title,
      size: "small",
      show,
      dismissible,
      onDismiss,
      footer: showFooter && /* @__PURE__ */ jsxs(Flex, { justifyContent: "space-between", width: "100%", children: [
        onReset && /* @__PURE__ */ jsxs(Fragment, { children: [
          /* @__PURE__ */ jsx(
            Button,
            {
              onClick: onReset,
              variant: "emphasized",
              disabled: disableButtons?.reset === true,
              children: controlLabels?.reset
            }
          ),
          /* @__PURE__ */ jsx(
            "div",
            {
              id: statusId,
              role: "status",
              "aria-live": "polite",
              "aria-relevant": "all",
              className: screenReaderOnly,
              children: showResetStatus && intl.formatMessage(dataTableMessages.resetSuccessful)
            }
          )
        ] }),
        /* @__PURE__ */ jsxs(Flex, { justifyContent: "flex-end", gap: 8, children: [
          onDismiss && /* @__PURE__ */ jsx(Button, { onClick: onDismiss, variant: "default", children: controlLabels?.cancel }),
          onApply && /* @__PURE__ */ jsx(
            Button,
            {
              onClick: onApply,
              variant: "accent",
              color: "primary",
              disabled: disableButtons?.apply === true,
              children: controlLabels?.apply
            }
          )
        ] })
      ] }),
      children
    }
  );
});
export {
  DataTableSettingsModal
};
