// @dynatrace/strato-components/esm/charts/core/components/canvas/Canvas.js
import { jsx } from "react/jsx-runtime";
import {
  memo,
  useContext,
  useEffect,
  useRef,
  useDeferredValue,
  forwardRef,
  useImperativeHandle
} from "react";
import { CanvasContext } from "./CanvasContext.js";
const setupCanvas = (canvasElement, width, height) => {
  const dpr = Math.max(1, window.devicePixelRatio) || 1;
  canvasElement.width = width * dpr;
  canvasElement.height = height * dpr;
  const context = canvasElement.getContext("2d");
  context?.scale(dpr, dpr);
  canvasElement.style.width = "100%";
  canvasElement.style.height = "100%";
  return context;
};
const draw = (canvasElement, context, shapeRefs) => {
  if (!canvasElement) {
    return;
  }
  if (context) {
    context.clearRect(0, 0, canvasElement.width, canvasElement.height);
    shapeRefs.forEach((ref) => {
      ref.current.draw(context, ref.current.id);
    });
  }
};
function ensureCanvasContext(currentCtx, canvasRef, width, height) {
  if (!currentCtx.ctx || currentCtx.prevHeight !== height || currentCtx.prevWidth !== width) {
    if (canvasRef.current) {
      currentCtx.ctx = setupCanvas(canvasRef.current, width, height);
    } else {
      currentCtx.ctx = null;
    }
    currentCtx.prevHeight = height;
    currentCtx.prevWidth = width;
  }
}
const MIN_WIDTH_CANVAS = 15;
const Canvas = memo(
  forwardRef(
    ({ children, ...props }, ref) => {
      const { height, width } = props;
      const canvasRef = useRef(null);
      const { shapeRefs } = useContext(CanvasContext);
      const contextRef = useRef({
        ctx: null,
        prevHeight: 0,
        prevWidth: 0
      });
      useImperativeHandle(
        ref,
        () => {
          return {
            customSetStyle: (key, value) => {
              canvasRef.current?.style.setProperty(key, value);
            }
          };
        },
        []
      );
      const isReady = useDeferredValue(width > MIN_WIDTH_CANVAS);
      useEffect(() => {
        if (!isReady || !canvasRef) {
          return;
        }
        ensureCanvasContext(contextRef.current, canvasRef, width, height);
        if (contextRef.current.ctx) {
          const shapes = [...shapeRefs.values()];
          requestAnimationFrame(() => {
            if (contextRef.current.ctx) {
              draw(canvasRef.current, contextRef.current.ctx, shapes);
            }
          });
        }
      });
      return /* @__PURE__ */ jsx("canvas", { ref: canvasRef, children });
    }
  )
);
export {
  Canvas
};
