// @dynatrace/strato-components/charts/single-value/utils/single-value-config.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var single_value_config_exports = {};
__export(single_value_config_exports, {
  buildSingleValueConfig: () => buildSingleValueConfig,
  buildSparklineConfig: () => buildSparklineConfig,
  deserializeConfig: () => deserializeConfig,
  isSerializedConfig: () => isSerializedConfig,
  normalizeConfig: () => normalizeConfig
});
module.exports = __toCommonJS(single_value_config_exports);
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_get_toolbar_config_without_intents = require("../../core/components/toolbar/utils/get-toolbar-config-without-intents.js");
var import_ColorRule = require("../../core/slots/ColorRule.js");
var import_defaults = require("../defaults.js");
var import_single_value_toolbar = require("../types/single-value-toolbar.js");
function getNodeInnerText(node) {
  if (typeof node === "string") {
    return node;
  }
  if (node === null || typeof node === "boolean" || typeof node === "undefined") {
    return "";
  }
  if (typeof node === "number") {
    return node.toString();
  }
  if ((0, import_react.isValidElement)(node) && node.props && node.props.children) {
    return getNodeInnerText(node.props.children);
  }
  return "";
}
function deserializeConfig(serializedConfig) {
  try {
    return (0, import_lodash_es.merge)({}, import_defaults.defaultSingleValueConfig, JSON.parse(serializedConfig));
  } catch {
    return import_defaults.defaultSingleValueConfig;
  }
}
function isSerializedConfig(contextConfig) {
  return typeof contextConfig === "string";
}
function normalizeConfig(config) {
  const sparklineConfig = config.sparkline;
  const toolbar = config.toolbar ? (0, import_get_toolbar_config_without_intents.getToolbarConfigWithoutIntents)(config.toolbar) : void 0;
  const output = {
    ...config,
    label: getNodeInnerText(config.label),
    toolbar
  };
  if (sparklineConfig) {
    output.sparkline = sparklineConfig;
  }
  return output;
}
const buildSparklineConfig = (slotsSparklineConfig, contextSparklineConfig) => {
  if (!slotsSparklineConfig) {
    return;
  }
  const contextThresholds = contextSparklineConfig?.thresholdIndicators || [];
  const slotThresholds = slotsSparklineConfig?.thresholdIndicators || [];
  return {
    thresholdIndicators: contextThresholds.concat(slotThresholds),
    color: slotsSparklineConfig?.color ?? contextSparklineConfig?.color,
    variant: slotsSparklineConfig?.variant ?? contextSparklineConfig?.variant,
    xAxis: {
      ...contextSparklineConfig?.xAxis,
      ...slotsSparklineConfig?.xAxis
    },
    yAxis: {
      ...contextSparklineConfig?.yAxis,
      ...slotsSparklineConfig?.yAxis
    },
    showTicks: slotsSparklineConfig?.showTicks ?? contextSparklineConfig?.showTicks,
    showContextValues: slotsSparklineConfig?.showContextValues ?? contextSparklineConfig?.showContextValues,
    gapPolicy: slotsSparklineConfig?.gapPolicy ?? contextSparklineConfig?.gapPolicy,
    curve: slotsSparklineConfig?.curve
  };
};
const getDefaultToolbarPlacement = (singleValueAlignment) => {
  return singleValueAlignment === "end" ? "top-left" : import_single_value_toolbar.defaultToolbarInternalConfig.placement;
};
const getToolbarConfig = (toolbarSlotsConfig, toolbarConsumerConfig, singleValueAlignment) => {
  if (!toolbarSlotsConfig && !toolbarConsumerConfig) {
    return void 0;
  }
  const placement = toolbarSlotsConfig?.placement ?? toolbarConsumerConfig?.placement ?? getDefaultToolbarPlacement(singleValueAlignment);
  return (0, import_lodash_es.merge)(
    {},
    import_single_value_toolbar.defaultToolbarInternalConfig,
    toolbarConsumerConfig,
    toolbarSlotsConfig,
    { placement }
  );
};
function buildSingleValueConfig(props, slots, contextConfig) {
  const colorRules = (contextConfig.colorRules || []).filter(import_ColorRule.isValidColorRule).concat(slots.colorRules || []);
  const toolbar = getToolbarConfig(
    slots.toolbar,
    contextConfig.toolbar,
    props.alignment
  );
  return {
    label: props.label ?? contextConfig.label,
    color: props.color ?? contextConfig.color,
    unit: props.unit ?? contextConfig.unit ?? import_defaults.defaultSingleValueConfig.unit,
    formatter: props.formatter ?? contextConfig.formatter,
    alignment: props.alignment ?? contextConfig.alignment ?? import_defaults.defaultSingleValueConfig.alignment,
    sparkline: buildSparklineConfig(slots.sparkline, contextConfig.sparkline),
    colorRules,
    toolbar
  };
}
