// @dynatrace/strato-components/buttons/intent-button/IntentButtonWithoutResponse.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var IntentButtonWithoutResponse_exports = {};
__export(IntentButtonWithoutResponse_exports, {
  IntentButtonWithoutResponse: () => IntentButtonWithoutResponse
});
module.exports = __toCommonJS(IntentButtonWithoutResponse_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_navigation = require("@dynatrace-sdk/navigation");
var styles = __toESM(require("./IntentButton.sty.js"));
var import_intent = require("../../core/hooks/intent.js");
var import_AppIcon = require("../../layouts/app-header/app-icon/AppIcon.js");
var import_Button = require("../button/Button.js");
const IntentButtonWithoutResponse = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    payload,
    options,
    icon,
    children,
    disabled,
    onClick,
    size,
    textAlign,
    "aria-disabled": ariaDisabledProp,
    ...remainingProps
  } = props;
  const intent = (0, import_intent.useIntent)({
    payload,
    options
  });
  const link = (0, import_react.useMemo)(() => {
    return (0, import_navigation.getIntentLink)(
      payload,
      options.recommendedAppId ?? "",
      options.recommendedIntentId ?? ""
    );
  }, [options.recommendedAppId, options.recommendedIntentId, payload]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_Button.Button,
    {
      as: "a",
      href: disabled ? void 0 : link,
      ref: forwardedRef,
      onClick: (event) => {
        onClick?.(event);
        if (event.isDefaultPrevented()) {
          return;
        }
        const nativeEvent = event.nativeEvent;
        if (nativeEvent.button !== 0 || nativeEvent.metaKey || nativeEvent.ctrlKey || nativeEvent.shiftKey) {
          return;
        }
        intent.send();
        event.stopPropagation();
        event.preventDefault();
      },
      onKeyDown: (event) => {
        const nativeEvent = event.nativeEvent;
        if (nativeEvent.code !== "Space" && nativeEvent.code !== "Enter") {
          return;
        }
        event.preventDefault();
        if (nativeEvent.metaKey || nativeEvent.ctrlKey) {
          window.open(link);
          return;
        }
        if (nativeEvent.shiftKey) {
          window.open(link, "_blank", "menubar");
          return;
        }
        intent.send();
      },
      disabled,
      "aria-disabled": ariaDisabledProp,
      textAlign,
      size,
      ...remainingProps,
      children: [
        intent.iconURL && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_AppIcon.AppIcon,
          {
            src: intent.iconURL,
            className: disabled || ariaDisabledProp ? styles.disabledAppIconOpacity : void 0,
            alt: ""
          }
        ) }),
        icon && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: icon }),
        children
      ]
    }
  );
});
