// @dynatrace/strato-components/buttons/intent-button/IntentButton.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var IntentButton_exports = {};
__export(IntentButton_exports, {
  IntentButton: () => IntentButton
});
module.exports = __toCommonJS(IntentButton_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react2 = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_IntentButtonWithoutResponse = require("./IntentButtonWithoutResponse.js");
var import_IntentButtonWithResponse = require("./IntentButtonWithResponse.js");
var import_OpenWith = require("./OpenWith.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useSlots = require("../../core/hooks/useSlots.js");
var import_Intent = require("../../core/slots/Intent/Intent.js");
var import_merge_props = require("../../core/utils/merge-props.js");
var import_InputGroup = require("../../layouts/input-group/InputGroup.js");
var import_Menu = require("../../navigation/menu/Menu.js");
var import_Prefix = require("../../navigation/menu/Prefix.js");
var import_Button = require("../button/Button.js");
const COMPONENT_NAME = "IntentButton";
const _IntentButton = /* @__PURE__ */ (0, import_react2.forwardRef)((props, forwardedRef) => {
  const {
    payload,
    children,
    onResponse,
    options,
    disabled,
    color,
    variant = "emphasized",
    width,
    onClick,
    size,
    textAlign,
    showMenu = true,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "aria-disabled": ariaDisabledProp,
    className: consumerClassName,
    style: consumerStyle,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const [slots, freeform] = (0, import_useSlots.useSlots)(children, {
    icon: import_Prefix.Prefix,
    items: { match: import_Intent.Intent, multiple: true },
    openWith: import_OpenWith.OpenWith
  });
  const slottedIcon = (0, import_react2.isValidElement)(slots.icon) ? slots.icon : void 0;
  const intl = (0, import_react_intl.useIntl)();
  const menuAriaLabel = intl.formatMessage({
    defaultMessage: "Open intent menu",
    id: "OEhnk3JokTy0c/JX",
    description: "Aria label for the IntentButton menu trigger."
  });
  const sharedProps = {
    color,
    variant,
    width,
    size,
    textAlign,
    onClick,
    disabled,
    "aria-disabled": ariaDisabledProp
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_InputGroup.InputGroup,
    {
      ref: forwardedRef,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      className: consumerClassName,
      style: consumerStyle,
      ...(0, import_merge_props.mergeProps)(restProps, behavioralTrackingProps),
      "data-dt-component": COMPONENT_NAME,
      children: [
        "responseProperties" in options ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_IntentButtonWithResponse.IntentButtonWithResponse,
          {
            ...sharedProps,
            payload,
            options,
            onResponse: onResponse ?? (() => void 0),
            icon: slottedIcon,
            children: freeform
          }
        ) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_IntentButtonWithoutResponse.IntentButtonWithoutResponse,
          {
            ...sharedProps,
            payload,
            options,
            icon: slottedIcon,
            children: freeform
          }
        ),
        showMenu && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Menu.Menu, { children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Trigger, { "aria-label": menuAriaLabel, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_Button.Button,
            {
              disabled,
              "aria-disabled": ariaDisabledProp,
              variant,
              color,
              size,
              style: { width: "max-content", minWidth: "fit-content" },
              children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.DotMenuIcon, {})
            }
          ) }),
          /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
            import_Menu.Menu.Content,
            {
              alignment: "end",
              "data-dtrum-mask": dataDtrumMask,
              "data-dtrum-allow": dataDtrumAllow,
              children: [
                slots.items && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Group, { children: slots.items.map((item, index) => /* @__PURE__ */ (0, import_react.createElement)(
                  import_Menu.Menu.Intent,
                  {
                    ...item.props,
                    key: `${item.props.options.recommendedAppId}-${index}`
                  }
                )) }),
                /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Group, { children: slots.openWith ?? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OpenWith.OpenWith, { payload }) })
              ]
            }
          )
        ] })
      ]
    }
  );
});
_IntentButton.displayName = COMPONENT_NAME;
const IntentButton = Object.assign(_IntentButton, {
  Item: import_Intent.Intent,
  OpenWith: import_OpenWith.OpenWith,
  Icon: import_Prefix.Prefix
});
