// @dynatrace/strato-components/buttons/intent-button/OpenWith.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var OpenWith_exports = {};
__export(OpenWith_exports, {
  OpenWith: () => OpenWith
});
module.exports = __toCommonJS(OpenWith_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_merge_props = require("../../core/utils/merge-props.js");
var import_Menu = require("../../navigation/menu/Menu.js");
const OpenWith = /* @__PURE__ */ (0, import_react.forwardRef)(({ ...allProps }, forwardedRef) => {
  const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(allProps);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Menu.Menu.Intent,
    {
      ...(0, import_merge_props.mergeProps)(restProps, behavioralTrackingProps),
      options: { keyProperties: [] },
      ref: forwardedRef,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_react_intl.FormattedMessage,
        {
          id: "F6H1FQ86mt53ZgVL",
          defaultMessage: "Open with...",
          description: "Default label for the intent button component."
        }
      )
    }
  );
});
OpenWith.displayName = "IntentButton.OpenWith";
