// @dynatrace/strato-components/forms/text-input/TextInput.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TextInput_exports = {};
__export(TextInput_exports, {
  TextInput: () => TextInput
});
module.exports = __toCommonJS(TextInput_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useControllableState = require("../../core/hooks/useControllableState.js");
var import_merge_props = require("../../core/utils/merge-props.js");
var import_BaseInput = require("../_base-input/BaseInput.js");
var import_BaseInputButton = require("../_base-input/BaseInputButton.js");
var import_BaseInputContent = require("../_base-input/BaseInputContent.js");
const COMPONENT_NAME = "TextInput";
const TextInputRoot = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    onChange,
    value,
    defaultValue,
    required,
    readOnly,
    type = "text",
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "aria-disabled": ariaDisabledProp,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const [internalValue, setInternalValue] = (0, import_useControllableState.useControllableState)({
    defaultValue,
    value
  });
  const handleOnChange = (e) => {
    onChange?.(e.target.value, e);
    setInternalValue(e.target.value);
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_BaseInput.BaseInput,
    {
      ref: forwardedRef,
      ...(0, import_merge_props.mergeProps)(restProps, behavioralTrackingProps),
      type,
      value: internalValue,
      required,
      readOnly,
      onChange: handleOnChange,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "aria-disabled": ariaDisabledProp,
      "data-dt-component": COMPONENT_NAME
    }
  );
});
const TextInput = Object.assign(TextInputRoot, {
  Prefix: import_BaseInputContent.BaseInputPrefix,
  Suffix: import_BaseInputContent.BaseInputSuffix,
  Button: import_BaseInputButton.BaseInputButton
});
TextInput.displayName = COMPONENT_NAME;
