// @dynatrace/strato-components/esm/charts/core/components/axes/utils/get-tick-label-props.js
import { getTickValueAsNumber } from "../../../utils/get-tick-value-as-number.js";
import {
  TICK_ALIGMENT,
  X_AXES_TICK_LINE_OFFSET,
  Y_AXES_TICK_LINE_OFFSET
} from "../constants.js";
const getBottomTickLabelPropsWithDomainAlignment = (domainMin, domainMax) => {
  return (value) => {
    const numericValue = getTickValueAsNumber(value);
    let textAnchor = TICK_ALIGMENT.Middle;
    if (numericValue === domainMin) {
      textAnchor = TICK_ALIGMENT.Start;
    } else if (numericValue === domainMax) {
      textAnchor = TICK_ALIGMENT.End;
    }
    return {
      y: X_AXES_TICK_LINE_OFFSET,
      textAnchor,
      verticalAnchor: TICK_ALIGMENT.Start
    };
  };
};
const VISX_BOTTOM_OFFSET = 2;
const ELLIPSIS_BUFFER = 2;
const leftTickLabelProps = (_value, index, values) => ({
  x: -Y_AXES_TICK_LINE_OFFSET,
  textAnchor: TICK_ALIGMENT.End,
  verticalAnchor: index === 0 ? TICK_ALIGMENT.End : index === values.length - 1 ? TICK_ALIGMENT.Start : TICK_ALIGMENT.Middle
});
const rightTickLabelProps = (_value, index, values) => ({
  x: +Y_AXES_TICK_LINE_OFFSET,
  textAnchor: TICK_ALIGMENT.Start,
  verticalAnchor: index === 0 ? TICK_ALIGMENT.End : index === values.length - 1 ? TICK_ALIGMENT.Start : TICK_ALIGMENT.Middle
});
const leftTickCategoricalLabelProps = (_value) => ({
  x: -Y_AXES_TICK_LINE_OFFSET,
  textAnchor: TICK_ALIGMENT.End,
  verticalAnchor: TICK_ALIGMENT.Middle
});
const topTickLabelProps = (_value, index, values) => ({
  y: -X_AXES_TICK_LINE_OFFSET,
  textAnchor: index === 0 ? TICK_ALIGMENT.Start : index === values.length - 1 ? TICK_ALIGMENT.End : TICK_ALIGMENT.Middle,
  verticalAnchor: TICK_ALIGMENT.End
});
const topRotatedTickLabelProps = (_value, index, values) => ({
  y: X_AXES_TICK_LINE_OFFSET,
  textAnchor: index === 0 ? TICK_ALIGMENT.Start : index === values.length - 1 ? TICK_ALIGMENT.End : TICK_ALIGMENT.Middle,
  verticalAnchor: TICK_ALIGMENT.Middle,
  angle: -90
});
const topTickLabelPropsForCategoryAxis = (_value) => ({
  y: X_AXES_TICK_LINE_OFFSET,
  textAnchor: TICK_ALIGMENT.Middle,
  verticalAnchor: TICK_ALIGMENT.End
});
const topRotatedTickLabelPropsForCategoryAxis = (_value) => ({
  y: X_AXES_TICK_LINE_OFFSET,
  textAnchor: TICK_ALIGMENT.End,
  verticalAnchor: TICK_ALIGMENT.Middle,
  angle: -90
});
const bottomTickLabelProps = (_value, index, values) => ({
  y: X_AXES_TICK_LINE_OFFSET,
  textAnchor: index === 0 ? TICK_ALIGMENT.Start : index === values.length - 1 ? TICK_ALIGMENT.End : TICK_ALIGMENT.Middle,
  verticalAnchor: TICK_ALIGMENT.Start
});
const bottomRotatedTickLabelPropsForCategoryAxis = (_value) => ({
  y: X_AXES_TICK_LINE_OFFSET,
  textAnchor: TICK_ALIGMENT.End,
  verticalAnchor: TICK_ALIGMENT.Middle,
  angle: -90
});
const bottomRotatedTickLabelProps = (_value, index, values) => ({
  y: X_AXES_TICK_LINE_OFFSET,
  textAnchor: index === 0 ? TICK_ALIGMENT.Start : index === values.length - 1 ? TICK_ALIGMENT.End : TICK_ALIGMENT.Middle,
  verticalAnchor: TICK_ALIGMENT.Middle,
  angle: -90
});
const bottomTickLabelPropsForCategoryAxis = (_value) => ({
  y: X_AXES_TICK_LINE_OFFSET,
  textAnchor: TICK_ALIGMENT.Middle,
  verticalAnchor: TICK_ALIGMENT.Start
});
const getTopTickLabelPropsWithDomainAlignment = (domainMin, domainMax) => {
  return (value) => {
    const numericValue = getTickValueAsNumber(value);
    let textAnchor = TICK_ALIGMENT.Middle;
    if (numericValue === domainMin) {
      textAnchor = TICK_ALIGMENT.Start;
    } else if (numericValue === domainMax) {
      textAnchor = TICK_ALIGMENT.End;
    }
    return {
      y: -X_AXES_TICK_LINE_OFFSET,
      textAnchor,
      verticalAnchor: TICK_ALIGMENT.End
    };
  };
};
const getTickLabelPropsForCategoryAxis = (orientation, tickLabelLayout = "horizontal") => {
  if (orientation === "left") {
    return leftTickCategoricalLabelProps;
  }
  if (orientation === "top") {
    return tickLabelLayout === "vertical" ? topRotatedTickLabelPropsForCategoryAxis : topTickLabelPropsForCategoryAxis;
  }
  return tickLabelLayout === "vertical" ? bottomRotatedTickLabelPropsForCategoryAxis : bottomTickLabelPropsForCategoryAxis;
};
const bottomTickLabelDateProps = (_value) => ({
  y: X_AXES_TICK_LINE_OFFSET + VISX_BOTTOM_OFFSET,
  textAnchor: TICK_ALIGMENT.Middle,
  verticalAnchor: TICK_ALIGMENT.Start
});
const bottomPanTickLabelDateProps = (_value, index, values) => {
  const result = {
    textAnchor: TICK_ALIGMENT.Middle,
    verticalAnchor: TICK_ALIGMENT.Middle
  };
  const isFirstItem = index === 0;
  const isLastItem = index === values.length - 1;
  if (isFirstItem) {
    result.textAnchor = TICK_ALIGMENT.Start;
  }
  if (isLastItem) {
    result.textAnchor = TICK_ALIGMENT.End;
  }
  return result;
};
const getTickLabelPropsForValueAxis = (orientation, tickLabelLayout = "horizontal") => {
  if (orientation === "left") {
    return leftTickLabelProps;
  }
  if (orientation === "right") {
    return rightTickLabelProps;
  }
  if (orientation === "top") {
    return tickLabelLayout === "vertical" ? topRotatedTickLabelProps : topTickLabelProps;
  }
  return tickLabelLayout === "vertical" ? bottomRotatedTickLabelProps : bottomTickLabelProps;
};
function getXAxisTickLabelProps(position, isCategoricalAxis, isTickRotated, scale, tickLabelLayout) {
  if (isCategoricalAxis) {
    return getTickLabelPropsForCategoryAxis(position, tickLabelLayout);
  }
  if (scale && !isTickRotated) {
    const domain = scale.domain();
    const getDomainAligned = position === "bottom" ? getBottomTickLabelPropsWithDomainAlignment : getTopTickLabelPropsWithDomainAlignment;
    return getDomainAligned(+domain[0], +domain[1]);
  }
  return getTickLabelPropsForValueAxis(position, tickLabelLayout);
}
const getTickLabelPropsForGanttAxis = (min, max) => {
  return (value, index, values) => {
    let textAnchor = TICK_ALIGMENT.Middle;
    let dx = 0;
    const valueAsNumber = getTickValueAsNumber(value);
    if (values.length === 1 && valueAsNumber && min && max) {
      textAnchor = valueAsNumber - min < valueAsNumber - max ? TICK_ALIGMENT.Start : TICK_ALIGMENT.End;
    } else if (index === 0) {
      textAnchor = TICK_ALIGMENT.Start;
      dx = valueAsNumber === min ? 4 : 0;
    } else if (index === values.length - 1) {
      textAnchor = TICK_ALIGMENT.End;
      dx = valueAsNumber === max ? -4 : 0;
    }
    return {
      dx,
      dy: -2,
      textAnchor,
      verticalAnchor: TICK_ALIGMENT.Middle
    };
  };
};
export {
  ELLIPSIS_BUFFER,
  bottomPanTickLabelDateProps,
  bottomTickLabelDateProps,
  getBottomTickLabelPropsWithDomainAlignment,
  getTickLabelPropsForCategoryAxis,
  getTickLabelPropsForGanttAxis,
  getTickLabelPropsForValueAxis,
  getTopTickLabelPropsWithDomainAlignment,
  getXAxisTickLabelProps
};
