// @dynatrace-sdk/client-davis-analyzers/cjs/index.js
/**
 * @license
 * Copyright 2023-2026 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/client/davis-analyzers/src/index.ts
var src_exports = {};
__export(src_exports, {
  AnalyzerDefinitionType: () => AnalyzerDefinitionType,
  AnalyzerDimensionalDataType: () => AnalyzerDimensionalDataType,
  AnalyzerErrorEnvelopeError: () => AnalyzerErrorEnvelopeError,
  AnalyzerExecutionLogLevel: () => AnalyzerExecutionLogLevel,
  AnalyzerOutputLogLevel: () => AnalyzerOutputLogLevel,
  AnalyzerResultExecutionStatus: () => AnalyzerResultExecutionStatus,
  AnalyzerResultResultStatus: () => AnalyzerResultResultStatus,
  AnalyzersClient: () => AnalyzersClient,
  ApiClientError: () => import_shared_errors4.ApiClientError,
  ApiGatewayError: () => import_shared_errors4.ApiGatewayError,
  ClientRequestError: () => import_shared_errors4.ClientRequestError,
  InvalidResponseError: () => import_shared_errors4.InvalidResponseError,
  JsonSchemaConfig: () => JsonSchemaConfig,
  ParameterDefinitionType: () => ParameterDefinitionType,
  analyzersClient: () => analyzersClient,
  isAnalyzerErrorEnvelopeError: () => isAnalyzerErrorEnvelopeError,
  isApiClientError: () => import_shared_errors4.isApiClientError,
  isApiGatewayError: () => import_shared_errors4.isApiGatewayError,
  isClientRequestError: () => import_shared_errors4.isClientRequestError,
  isInvalidResponseError: () => import_shared_errors4.isInvalidResponseError
});
module.exports = __toCommonJS(src_exports);

// packages/client/davis-analyzers/src/lib/apis/analyzers-api.ts
var import_http_client = require("@dynatrace-sdk/http-client");
var import_shared_errors3 = require("@dynatrace-sdk/shared-errors");

// packages/client/davis-analyzers/src/lib/error-envelopes/analyzer-error-envelope-error.ts
var import_shared_errors = require("@dynatrace-sdk/shared-errors");
var AnalyzerErrorEnvelopeError = class extends import_shared_errors.ClientRequestError {
  isAnalyzerErrorEnvelopeError = true;
};
function isAnalyzerErrorEnvelopeError(e) {
  return e?.isApiClientError === true && e?.isClientRequestError === true && e?.isAnalyzerErrorEnvelopeError === true && e instanceof Error;
}

// packages/client/davis-analyzers/src/lib/error-envelopes/api-gateway-errors-handler.ts
var import_shared_errors2 = require("@dynatrace-sdk/shared-errors");
async function apiGatewayErrorsHandler(response) {
  if (response.headers?.["dynatrace-response-source"] === "API Gateway") {
    const responseBody = await response.body("json");
    throw new import_shared_errors2.ApiGatewayError(response, responseBody);
  }
}

// packages/client/davis-analyzers/src/lib/error-envelopes/get-error-message.ts
function serializeData(data) {
  try {
    return JSON.stringify(data);
  } catch (e) {
    return String(data);
  }
}
function getMessagesFromErrorDetails(details) {
  const messages = [];
  Object.entries(details).forEach(([name, data]) => {
    if (!data)
      return;
    const serializedData = serializeData(data);
    switch (name) {
      case "missingScopes":
        messages.push(`Missing scopes: ${serializedData}`);
        break;
      default:
        messages.push(`${name}: ${serializedData}`);
    }
  }, []);
  return messages;
}
function getErrorMessage(errorBody, defaultMessage) {
  const error = errorBody;
  const msg = error?.error?.message || defaultMessage;
  const details = error?.error?.details || {};
  return [msg, ...getMessagesFromErrorDetails(details)].join(". ");
}

// packages/client/davis-analyzers/src/lib/utils/transformations.ts
var defaultRuleConfig = {
  force: false,
  datetime: false
};
function mergeKeys(key1, key2) {
  if (!key1)
    return key2;
  if (!key2)
    return key1;
  return `${key1}.${key2}`;
}
function matchKey(key, rules, cache) {
  if (cache.has(key)) {
    return cache.get(key);
  }
  for (const rule of rules) {
    const pattern = Array.isArray(rule) ? rule[0] : rule;
    const config = Array.isArray(rule) ? rule[1] : defaultRuleConfig;
    if (key === pattern) {
      const result = [pattern, config];
      cache.set(key, result);
      return result;
    }
    let i = 0;
    let j = 0;
    const keyLen = key.length;
    const patternLen = pattern.length;
    let isMatch = true;
    while (i < keyLen && j < patternLen) {
      let keyPartEnd = key.indexOf(".", i);
      let patternPartEnd = pattern.indexOf(".", j);
      if (keyPartEnd === -1)
        keyPartEnd = keyLen;
      if (patternPartEnd === -1)
        patternPartEnd = patternLen;
      const keyPart = key.slice(i, keyPartEnd);
      const patternPart = pattern.slice(j, patternPartEnd);
      if (patternPart !== "*" && patternPart !== "**" && keyPart !== patternPart) {
        isMatch = false;
        break;
      }
      i = keyPartEnd + 1;
      j = patternPartEnd + 1;
    }
    if (isMatch) {
      if (j < patternLen) {
        const remaining = pattern.slice(j);
        if (remaining !== "**" && remaining !== "*") {
          isMatch = false;
        }
      }
      if (i < keyLen && pattern.slice(j - 1) !== "**") {
        isMatch = false;
      }
    }
    if (isMatch) {
      const result = [pattern, config];
      cache.set(key, result);
      return result;
    }
  }
  cache.set(key, null);
  return null;
}
var potentialMatchForRoot = (pattern, key) => pattern.startsWith(key) || pattern.startsWith("*") || pattern.startsWith("**");
var potentialMatch = (pattern, key) => pattern.startsWith(key) || pattern.includes("*") || pattern.includes("**");
function partialMatch(key, keyPatterns) {
  const keyPatternsLen = keyPatterns.length;
  if (!key.includes(".")) {
    for (let i = 0; i < keyPatternsLen; i++) {
      const pattern = keyPatterns[i];
      if (potentialMatchForRoot(pattern, key))
        return true;
    }
    return false;
  }
  let dotIndex = -1;
  let currentKeyEnd = 0;
  while ((dotIndex = key.indexOf(".", currentKeyEnd)) !== -1) {
    const currentKey = key.slice(0, dotIndex);
    const isRoot = !currentKey.includes(".");
    for (let i = 0; i < keyPatternsLen; i++) {
      const pattern = keyPatterns[i];
      if (isRoot) {
        if (potentialMatchForRoot(pattern, currentKey))
          return true;
      } else if (potentialMatch(pattern, currentKey))
        return true;
    }
    currentKeyEnd = dotIndex + 1;
  }
  for (let i = 0; i < keyPatternsLen; i++) {
    const pattern = keyPatterns[i];
    if (potentialMatch(pattern, key))
      return true;
  }
  return false;
}
var dateTimePattern = /^(\d{4}-\d{2}-\d{2})(T\d{2}:\d{2}:\d{2}(?:\.\d+)?(Z|[+-]\d{2}:\d{2})?)?$/i;
function isDateLike(value) {
  if (typeof value === "string") {
    return dateTimePattern.test(value);
  }
  return false;
}
function transformValue(value, rules, direction) {
  const { force, datetime } = rules;
  if (direction === "to") {
    return force || isDateLike(value) ? new Date(value) : value;
  }
  if (value instanceof Date) {
    const iso = value.toISOString();
    return datetime ? iso : iso.split("T")[0];
  }
  return value;
}
function walk(obj, direction, rules, rulesCache, keyPatterns, currentKey = "") {
  if (rules.length <= 0) {
    return obj;
  }
  if (!partialMatch(currentKey, keyPatterns)) {
    return obj;
  }
  if (Array.isArray(obj)) {
    const match = matchKey(currentKey, rules, rulesCache);
    obj.forEach((item, idx) => {
      if (typeof item === "object" && item !== null) {
        walk(item, direction, rules, rulesCache, keyPatterns, currentKey);
      } else if (match) {
        obj[idx] = transformValue(obj[idx], match[1], direction);
      }
    });
  }
  if (typeof obj === "object" && obj !== null && !Array.isArray(obj)) {
    for (const key of Object.keys(obj)) {
      const mergedKey = mergeKeys(currentKey, key);
      const match = matchKey(mergedKey, rules, rulesCache);
      if (match) {
        obj[key] = transformValue(obj[key], match[1], direction);
      }
      walk(obj[key], direction, rules, rulesCache, keyPatterns, mergedKey);
    }
  }
  return obj;
}
function flattenKeyPatterns(keyPatterns) {
  return keyPatterns.map((k) => Array.isArray(k) ? k[0] : k);
}
function transform(direction, object, rules) {
  const flatKeyPatterns = flattenKeyPatterns(rules);
  const rulesCache = /* @__PURE__ */ new Map();
  return walk(object, direction, rules, rulesCache, flatKeyPatterns);
}
function transformRequest(object, rules) {
  return transform("from", object, rules);
}
function transformResponse(object, rules) {
  return transform("to", object, rules);
}

// packages/client/davis-analyzers/src/lib/utils/url-helpers.ts
var encodeQueryParam = (key, value) => {
  const encodedKey = encodeURIComponent(key);
  return `${encodedKey}=${encodeURIComponent(typeof value === "number" ? value : String(value))}`;
};
var addExplodedArrayQueryParam = (query, key) => {
  const arrayValue = query[key];
  return arrayValue.map((value) => encodeQueryParam(key, value)).join("&");
};
var addNonExplodedArrayQueryParams = (query, key) => {
  const encodedKey = encodeURIComponent(key);
  const encodedParamsList = query[key].map((value) => encodeURIComponent(typeof value === "number" ? value : String(value))).join(",");
  return `${encodedKey}=${encodedParamsList}`;
};
var addQueryParam = (query, key) => encodeQueryParam(key, query[key]);
var arrayQueryParams = (query, key, explode) => {
  return explode ? addExplodedArrayQueryParam(query, key) : addNonExplodedArrayQueryParams(query, key);
};
var toQueryString = (rawQuery, flags = {}) => {
  const query = rawQuery || {};
  const keys = Object.keys(query).filter((key) => typeof query[key] !== "undefined");
  const queryString = keys.map(
    (key) => Array.isArray(query[key]) ? arrayQueryParams(query, key, flags.explode?.hasOwnProperty(key) ? flags.explode[key] : true) : addQueryParam(query, key)
  ).join("&");
  return queryString ? `?${queryString}` : "";
};

// packages/client/davis-analyzers/src/lib/apis/analyzers-api.ts
var AnalyzersClient = class {
  httpClient;
  shouldTransformDates = true;
  /**
   * @param {HttpClient} httpClientImplementation - You can provide custom http client as a parameter to constructor of a AnalyzersClient. Custom http client should implement {HttpClient} interface.
   * @param {ApiClientOptions} [options] - Optional configuration for SDK Client.
   * @example
   * const analyzersClientNodeJs = new AnalyzersClient(yourCustomImplementation);
   */
  constructor(httpClientImplementation, options) {
    this.httpClient = httpClientImplementation;
    this.shouldTransformDates = options?.transformDates ?? true;
  }
  /**
   * Query all available analyzer definitions.
   *
   * **Required scope:** davis:analyzers:read<br/><br/>The analyzer definition contains meta-information about an analyzer, including its input and output structure:
   * * The `input` definition specifies which fields the analyzer accepts, some of which may be required.
   * * The `output` definition specifies how the returned analyzer result is structured.
   *
   * @returns {Promise<AnalyzerQueryResult>} All available analyzer definitions.
   *
   * @example <caption>Code example</caption>
   * import { analyzersClient } from "@dynatrace-sdk/client-davis-analyzers";
   *
   * const data = await analyzersClient.queryAnalyzers();
   *
   * @throws {AnalyzerErrorEnvelopeError} Insufficient permissions. | Client error. | Internal server error.
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async queryAnalyzers(config = {}) {
    if (!config) {
      throw new import_shared_errors3.ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const query = toQueryString({
      "page-key": config.pageKey,
      "page-size": config.pageSize,
      filter: config.filter,
      "add-fields": config.addFields
    });
    try {
      const response = await this.httpClient.send({
        url: `/platform/davis/analyzers/v1/analyzers${query}`,
        method: "GET",
        headers: {
          Accept: "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
      } catch (err) {
        throw new import_shared_errors3.InvalidResponseError(
          `AnalyzersClient.queryAnalyzers:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if ((0, import_http_client.isHttpClientAbortError)(e) || (0, import_shared_errors3.isInvalidResponseError)(e)) {
        throw e;
      }
      if (!(0, import_http_client.isHttpClientResponseError)(e)) {
        throw new import_shared_errors3.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      switch (response.status) {
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AnalyzerErrorEnvelopeError(
              `403`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Insufficient permissions."),
              e
            );
          } catch (err) {
            if (err instanceof AnalyzerErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `AnalyzersClient.queryAnalyzers:403`,
              err,
              responseValue,
              "AnalyzerErrorEnvelope",
              void 0,
              response
            );
          }
        }
        default: {
          if (response.status >= 400 && response.status <= 499) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
              throw new AnalyzerErrorEnvelopeError(
                `${response.status}`,
                response,
                errorBody,
                getErrorMessage(errorBody, "Client error."),
                e
              );
            } catch (err) {
              if (err instanceof AnalyzerErrorEnvelopeError) {
                throw err;
              }
              throw new import_shared_errors3.InvalidResponseError(
                `AnalyzersClient.queryAnalyzers:${response.status}`,
                err,
                responseValue,
                "AnalyzerErrorEnvelope",
                void 0,
                response
              );
            }
          } else if (response.status >= 500 && response.status <= 599) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
              throw new AnalyzerErrorEnvelopeError(
                `${response.status}`,
                response,
                errorBody,
                getErrorMessage(errorBody, "Internal server error."),
                e
              );
            } catch (err) {
              if (err instanceof AnalyzerErrorEnvelopeError) {
                throw err;
              }
              throw new import_shared_errors3.InvalidResponseError(
                `AnalyzersClient.queryAnalyzers:${response.status}`,
                err,
                responseValue,
                "AnalyzerErrorEnvelope",
                void 0,
                response
              );
            }
          } else {
            try {
              const responseValue = await response.body("json");
              throw new import_shared_errors3.ClientRequestError(
                `${response.status}`,
                response,
                responseValue,
                getErrorMessage(
                  responseValue,
                  `Unexpected api response: code=${response.status} body="${responseValue}"`
                )
              );
            } catch (error) {
              if (error instanceof import_shared_errors3.ClientRequestError) {
                throw error;
              }
              const responseValue = await response.body("text").catch(() => "");
              throw new import_shared_errors3.InvalidResponseError(
                `${response.status}`,
                error,
                responseValue,
                "json",
                "Unexpected api response",
                response
              );
            }
          }
        }
      }
    }
  }
  /**
   * Get the analyzer definition for an analyzer.
   *
   * **Required scope:** davis:analyzers:read<br/><br/>All available meta-information for the analyzer is returned.
   *
   * @returns {Promise<AnalyzerDefinitionDetails>} The definition of the requested analyzer.
   *
   * @example <caption>Code example</caption>
   * import { analyzersClient } from "@dynatrace-sdk/client-davis-analyzers";
   *
   * const data = await analyzersClient.getAnalyzer({
   *   analyzerName: "...",
   * });
   *
   * @throws {AnalyzerErrorEnvelopeError} Invalid or malformed request. | Insufficient permissions. | No analyzer available with the requested name. | Client error. | Internal server error.
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async getAnalyzer(config) {
    try {
      const response = await this.httpClient.send({
        url: `/platform/davis/analyzers/v1/analyzers/${config.analyzerName}`,
        method: "GET",
        headers: {
          Accept: "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
      } catch (err) {
        throw new import_shared_errors3.InvalidResponseError(
          `AnalyzersClient.getAnalyzer:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if ((0, import_http_client.isHttpClientAbortError)(e) || (0, import_shared_errors3.isInvalidResponseError)(e)) {
        throw e;
      }
      if (!(0, import_http_client.isHttpClientResponseError)(e)) {
        throw new import_shared_errors3.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AnalyzerErrorEnvelopeError(
              `400`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Invalid or malformed request."),
              e
            );
          } catch (err) {
            if (err instanceof AnalyzerErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `AnalyzersClient.getAnalyzer:400`,
              err,
              responseValue,
              "AnalyzerErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AnalyzerErrorEnvelopeError(
              `403`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Insufficient permissions."),
              e
            );
          } catch (err) {
            if (err instanceof AnalyzerErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `AnalyzersClient.getAnalyzer:403`,
              err,
              responseValue,
              "AnalyzerErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AnalyzerErrorEnvelopeError(
              `404`,
              response,
              errorBody,
              getErrorMessage(errorBody, "No analyzer available with the requested name."),
              e
            );
          } catch (err) {
            if (err instanceof AnalyzerErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `AnalyzersClient.getAnalyzer:404`,
              err,
              responseValue,
              "AnalyzerErrorEnvelope",
              void 0,
              response
            );
          }
        }
        default: {
          if (response.status >= 400 && response.status <= 499) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
              throw new AnalyzerErrorEnvelopeError(
                `${response.status}`,
                response,
                errorBody,
                getErrorMessage(errorBody, "Client error."),
                e
              );
            } catch (err) {
              if (err instanceof AnalyzerErrorEnvelopeError) {
                throw err;
              }
              throw new import_shared_errors3.InvalidResponseError(
                `AnalyzersClient.getAnalyzer:${response.status}`,
                err,
                responseValue,
                "AnalyzerErrorEnvelope",
                void 0,
                response
              );
            }
          } else if (response.status >= 500 && response.status <= 599) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
              throw new AnalyzerErrorEnvelopeError(
                `${response.status}`,
                response,
                errorBody,
                getErrorMessage(errorBody, "Internal server error."),
                e
              );
            } catch (err) {
              if (err instanceof AnalyzerErrorEnvelopeError) {
                throw err;
              }
              throw new import_shared_errors3.InvalidResponseError(
                `AnalyzersClient.getAnalyzer:${response.status}`,
                err,
                responseValue,
                "AnalyzerErrorEnvelope",
                void 0,
                response
              );
            }
          } else {
            try {
              const responseValue = await response.body("json");
              throw new import_shared_errors3.ClientRequestError(
                `${response.status}`,
                response,
                responseValue,
                getErrorMessage(
                  responseValue,
                  `Unexpected api response: code=${response.status} body="${responseValue}"`
                )
              );
            } catch (error) {
              if (error instanceof import_shared_errors3.ClientRequestError) {
                throw error;
              }
              const responseValue = await response.body("text").catch(() => "");
              throw new import_shared_errors3.InvalidResponseError(
                `${response.status}`,
                error,
                responseValue,
                "json",
                "Unexpected api response",
                response
              );
            }
          }
        }
      }
    }
  }
  /**
   * Get the documentation for an analyzer.
   *
   * **Required scope:** davis:analyzers:read<br/><br/>A documentation may provide further information and context for a given analyzer.
   * Not all analyzers provide documentation.
   *
   * The documentation is returned as Markdown.
   *
   * @returns {Promise<Binary>} The documentation of the requested analyzer.
   *
   * @example <caption>Code example</caption>
   * import { analyzersClient } from "@dynatrace-sdk/client-davis-analyzers";
   *
   * const data = await analyzersClient.getAnalyzerDocumentation(
   *   { analyzerName: "..." },
   * );
   *
   * @throws {AnalyzerErrorEnvelopeError} Invalid or malformed request. | Insufficient permissions. | No analyzer available with the requested name. | Client error. | Internal server error.
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async getAnalyzerDocumentation(config) {
    try {
      const response = await this.httpClient.send({
        url: `/platform/davis/analyzers/v1/analyzers/${config.analyzerName}/documentation`,
        method: "GET",
        headers: {
          Accept: "text/markdown; charset=utf-8, application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("binary");
      try {
        return responseValue;
      } catch (err) {
        throw new import_shared_errors3.InvalidResponseError(
          `AnalyzersClient.getAnalyzerDocumentation:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if ((0, import_http_client.isHttpClientAbortError)(e) || (0, import_shared_errors3.isInvalidResponseError)(e)) {
        throw e;
      }
      if (!(0, import_http_client.isHttpClientResponseError)(e)) {
        throw new import_shared_errors3.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AnalyzerErrorEnvelopeError(
              `400`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Invalid or malformed request."),
              e
            );
          } catch (err) {
            if (err instanceof AnalyzerErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `AnalyzersClient.getAnalyzerDocumentation:400`,
              err,
              responseValue,
              "AnalyzerErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AnalyzerErrorEnvelopeError(
              `403`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Insufficient permissions."),
              e
            );
          } catch (err) {
            if (err instanceof AnalyzerErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `AnalyzersClient.getAnalyzerDocumentation:403`,
              err,
              responseValue,
              "AnalyzerErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AnalyzerErrorEnvelopeError(
              `404`,
              response,
              errorBody,
              getErrorMessage(errorBody, "No analyzer available with the requested name."),
              e
            );
          } catch (err) {
            if (err instanceof AnalyzerErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `AnalyzersClient.getAnalyzerDocumentation:404`,
              err,
              responseValue,
              "AnalyzerErrorEnvelope",
              void 0,
              response
            );
          }
        }
        default: {
          if (response.status >= 400 && response.status <= 499) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
              throw new AnalyzerErrorEnvelopeError(
                `${response.status}`,
                response,
                errorBody,
                getErrorMessage(errorBody, "Client error."),
                e
              );
            } catch (err) {
              if (err instanceof AnalyzerErrorEnvelopeError) {
                throw err;
              }
              throw new import_shared_errors3.InvalidResponseError(
                `AnalyzersClient.getAnalyzerDocumentation:${response.status}`,
                err,
                responseValue,
                "AnalyzerErrorEnvelope",
                void 0,
                response
              );
            }
          } else if (response.status >= 500 && response.status <= 599) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
              throw new AnalyzerErrorEnvelopeError(
                `${response.status}`,
                response,
                errorBody,
                getErrorMessage(errorBody, "Internal server error."),
                e
              );
            } catch (err) {
              if (err instanceof AnalyzerErrorEnvelopeError) {
                throw err;
              }
              throw new import_shared_errors3.InvalidResponseError(
                `AnalyzersClient.getAnalyzerDocumentation:${response.status}`,
                err,
                responseValue,
                "AnalyzerErrorEnvelope",
                void 0,
                response
              );
            }
          } else {
            try {
              const responseValue = await response.body("json");
              throw new import_shared_errors3.ClientRequestError(
                `${response.status}`,
                response,
                responseValue,
                getErrorMessage(
                  responseValue,
                  `Unexpected api response: code=${response.status} body="${responseValue}"`
                )
              );
            } catch (error) {
              if (error instanceof import_shared_errors3.ClientRequestError) {
                throw error;
              }
              const responseValue = await response.body("text").catch(() => "");
              throw new import_shared_errors3.InvalidResponseError(
                `${response.status}`,
                error,
                responseValue,
                "json",
                "Unexpected api response",
                response
              );
            }
          }
        }
      }
    }
  }
  /**
   * Get the JSON schema for an analyzer input.
   *
   * **Required scope:** davis:analyzers:read<br/><br/>The [JSON schema](https://json-schema.org/) defines a standardized declaration of the input structure,
   * providing well-defined input documentation that is both human- and machine-readable.
   *
   * @returns {Promise<Record<string, any>>} The [JSON input schema](https://json-schema.org/) of the requested analyzer.
   *
   * @example <caption>Code example</caption>
   * import { analyzersClient } from "@dynatrace-sdk/client-davis-analyzers";
   *
   * const data = await analyzersClient.getJsonSchemaForInput({
   *   analyzerName: "...",
   * });
   *
   * @throws {AnalyzerErrorEnvelopeError} Invalid or malformed request. | Insufficient permissions. | No analyzer available with the requested name. | Client error. | Internal server error.
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async getJsonSchemaForInput(config) {
    if (!config) {
      throw new import_shared_errors3.ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const query = toQueryString({ config: config.config });
    try {
      const response = await this.httpClient.send({
        url: `/platform/davis/analyzers/v1/analyzers/${config.analyzerName}/json-schema/input${query}`,
        method: "GET",
        headers: {
          Accept: "application/schema+json, application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body();
      try {
        return responseValue;
      } catch (err) {
        throw new import_shared_errors3.InvalidResponseError(
          `AnalyzersClient.getJsonSchemaForInput:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if ((0, import_http_client.isHttpClientAbortError)(e) || (0, import_shared_errors3.isInvalidResponseError)(e)) {
        throw e;
      }
      if (!(0, import_http_client.isHttpClientResponseError)(e)) {
        throw new import_shared_errors3.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AnalyzerErrorEnvelopeError(
              `400`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Invalid or malformed request."),
              e
            );
          } catch (err) {
            if (err instanceof AnalyzerErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `AnalyzersClient.getJsonSchemaForInput:400`,
              err,
              responseValue,
              "AnalyzerErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AnalyzerErrorEnvelopeError(
              `403`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Insufficient permissions."),
              e
            );
          } catch (err) {
            if (err instanceof AnalyzerErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `AnalyzersClient.getJsonSchemaForInput:403`,
              err,
              responseValue,
              "AnalyzerErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AnalyzerErrorEnvelopeError(
              `404`,
              response,
              errorBody,
              getErrorMessage(errorBody, "No analyzer available with the requested name."),
              e
            );
          } catch (err) {
            if (err instanceof AnalyzerErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `AnalyzersClient.getJsonSchemaForInput:404`,
              err,
              responseValue,
              "AnalyzerErrorEnvelope",
              void 0,
              response
            );
          }
        }
        default: {
          if (response.status >= 400 && response.status <= 499) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
              throw new AnalyzerErrorEnvelopeError(
                `${response.status}`,
                response,
                errorBody,
                getErrorMessage(errorBody, "Client error."),
                e
              );
            } catch (err) {
              if (err instanceof AnalyzerErrorEnvelopeError) {
                throw err;
              }
              throw new import_shared_errors3.InvalidResponseError(
                `AnalyzersClient.getJsonSchemaForInput:${response.status}`,
                err,
                responseValue,
                "AnalyzerErrorEnvelope",
                void 0,
                response
              );
            }
          } else if (response.status >= 500 && response.status <= 599) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
              throw new AnalyzerErrorEnvelopeError(
                `${response.status}`,
                response,
                errorBody,
                getErrorMessage(errorBody, "Internal server error."),
                e
              );
            } catch (err) {
              if (err instanceof AnalyzerErrorEnvelopeError) {
                throw err;
              }
              throw new import_shared_errors3.InvalidResponseError(
                `AnalyzersClient.getJsonSchemaForInput:${response.status}`,
                err,
                responseValue,
                "AnalyzerErrorEnvelope",
                void 0,
                response
              );
            }
          } else {
            try {
              const responseValue = await response.body("json");
              throw new import_shared_errors3.ClientRequestError(
                `${response.status}`,
                response,
                responseValue,
                getErrorMessage(
                  responseValue,
                  `Unexpected api response: code=${response.status} body="${responseValue}"`
                )
              );
            } catch (error) {
              if (error instanceof import_shared_errors3.ClientRequestError) {
                throw error;
              }
              const responseValue = await response.body("text").catch(() => "");
              throw new import_shared_errors3.InvalidResponseError(
                `${response.status}`,
                error,
                responseValue,
                "json",
                "Unexpected api response",
                response
              );
            }
          }
        }
      }
    }
  }
  /**
   * Get the JSON schema for an analyzer result.
   *
   * **Required scope:** davis:analyzers:read<br/><br/>The [JSON schema](https://json-schema.org/) defines a standardized declaration of the output structure,
   * providing well-defined output documentation that is both human- and machine-readable.
   *
   * @returns {Promise<Record<string, any>>} The [JSON result schema](https://json-schema.org/) of the requested analyzer.
   *
   * @example <caption>Code example</caption>
   * import { analyzersClient } from "@dynatrace-sdk/client-davis-analyzers";
   *
   * const data = await analyzersClient.getJsonSchemaForResult({
   *   analyzerName: "...",
   * });
   *
   * @throws {AnalyzerErrorEnvelopeError} Invalid or malformed request. | Insufficient permissions. | No analyzer available with the requested name. | Client error. | Internal server error.
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async getJsonSchemaForResult(config) {
    if (!config) {
      throw new import_shared_errors3.ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const query = toQueryString({ config: config.config });
    try {
      const response = await this.httpClient.send({
        url: `/platform/davis/analyzers/v1/analyzers/${config.analyzerName}/json-schema/result${query}`,
        method: "GET",
        headers: {
          Accept: "application/schema+json, application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body();
      try {
        return responseValue;
      } catch (err) {
        throw new import_shared_errors3.InvalidResponseError(
          `AnalyzersClient.getJsonSchemaForResult:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if ((0, import_http_client.isHttpClientAbortError)(e) || (0, import_shared_errors3.isInvalidResponseError)(e)) {
        throw e;
      }
      if (!(0, import_http_client.isHttpClientResponseError)(e)) {
        throw new import_shared_errors3.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AnalyzerErrorEnvelopeError(
              `400`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Invalid or malformed request."),
              e
            );
          } catch (err) {
            if (err instanceof AnalyzerErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `AnalyzersClient.getJsonSchemaForResult:400`,
              err,
              responseValue,
              "AnalyzerErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AnalyzerErrorEnvelopeError(
              `403`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Insufficient permissions."),
              e
            );
          } catch (err) {
            if (err instanceof AnalyzerErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `AnalyzersClient.getJsonSchemaForResult:403`,
              err,
              responseValue,
              "AnalyzerErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AnalyzerErrorEnvelopeError(
              `404`,
              response,
              errorBody,
              getErrorMessage(errorBody, "No analyzer available with the requested name."),
              e
            );
          } catch (err) {
            if (err instanceof AnalyzerErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `AnalyzersClient.getJsonSchemaForResult:404`,
              err,
              responseValue,
              "AnalyzerErrorEnvelope",
              void 0,
              response
            );
          }
        }
        default: {
          if (response.status >= 400 && response.status <= 499) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
              throw new AnalyzerErrorEnvelopeError(
                `${response.status}`,
                response,
                errorBody,
                getErrorMessage(errorBody, "Client error."),
                e
              );
            } catch (err) {
              if (err instanceof AnalyzerErrorEnvelopeError) {
                throw err;
              }
              throw new import_shared_errors3.InvalidResponseError(
                `AnalyzersClient.getJsonSchemaForResult:${response.status}`,
                err,
                responseValue,
                "AnalyzerErrorEnvelope",
                void 0,
                response
              );
            }
          } else if (response.status >= 500 && response.status <= 599) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
              throw new AnalyzerErrorEnvelopeError(
                `${response.status}`,
                response,
                errorBody,
                getErrorMessage(errorBody, "Internal server error."),
                e
              );
            } catch (err) {
              if (err instanceof AnalyzerErrorEnvelopeError) {
                throw err;
              }
              throw new import_shared_errors3.InvalidResponseError(
                `AnalyzersClient.getJsonSchemaForResult:${response.status}`,
                err,
                responseValue,
                "AnalyzerErrorEnvelope",
                void 0,
                response
              );
            }
          } else {
            try {
              const responseValue = await response.body("json");
              throw new import_shared_errors3.ClientRequestError(
                `${response.status}`,
                response,
                responseValue,
                getErrorMessage(
                  responseValue,
                  `Unexpected api response: code=${response.status} body="${responseValue}"`
                )
              );
            } catch (error) {
              if (error instanceof import_shared_errors3.ClientRequestError) {
                throw error;
              }
              const responseValue = await response.body("text").catch(() => "");
              throw new import_shared_errors3.InvalidResponseError(
                `${response.status}`,
                error,
                responseValue,
                "json",
                "Unexpected api response",
                response
              );
            }
          }
        }
      }
    }
  }
  /**
   * Execute an analyzer with the given input.
   *
   * **Required scope:** davis:analyzers:execute<br/><br/>The analyzer execution is started asynchronously.
   * If the result is not available after the specified timeout, a request token is returned
   * to poll for the result. Otherwise, the final result is directly available and no request token is returned.
   *
   * Depending on the chosen analyzer and the analyzer input, additional scopes might be required, e.g.
   * `storage:buckets:read` and `storage:metrics:read` to enable the analyzer to read time series data from Grail™.
   *
   * @returns {Promise<AnalyzerExecuteResult>} The final analyzer result and execution status. | The current analyzer execution status and a request token and time-to-live (TTL) to continue polling the result with.
   *
   * @example <caption>Code example</caption>
   * import { analyzersClient } from "@dynatrace-sdk/client-davis-analyzers";
   *
   * const data = await analyzersClient.executeAnalyzer({
   *   analyzerName: "...",
   *   body: {
   *     timeSeriesData: {
   *       expression: "timeseries avg(dt.host.cpu.usage)",
   *     },
   *     forecastHorizon: 10,
   *   },
   * });
   *
   * @throws {AnalyzerErrorEnvelopeError} Invalid or malformed request. | Insufficient permissions. | No analyzer available with the requested name. | Client error. | Internal server error.
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async executeAnalyzer(config) {
    if (!config) {
      throw new import_shared_errors3.ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const encodedBody = this.shouldTransformDates ? transformRequest(config.body, []) : config.body;
    const query = toQueryString({ "timeout-seconds": config.timeoutSeconds, "enable-preview": config.enablePreview });
    try {
      const response = await this.httpClient.send({
        url: `/platform/davis/analyzers/v1/analyzers/${config.analyzerName}:execute${query}`,
        method: "POST",
        requestBodyType: "json",
        body: encodedBody,
        headers: {
          "Content-Type": "application/json",
          Accept: "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200, 202].includes(status);
        }
      });
      switch (response.status) {
        case 200: {
          const responseValue = await response.body("json");
          try {
            return this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
          } catch (err) {
            throw new import_shared_errors3.InvalidResponseError(
              `AnalyzersClient.executeAnalyzer:${response.status}`,
              err,
              responseValue,
              void 0,
              void 0,
              void 0
            );
          }
        }
        case 202: {
          const responseValue = await response.body("json");
          try {
            return this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
          } catch (err) {
            throw new import_shared_errors3.InvalidResponseError(
              `AnalyzersClient.executeAnalyzer:${response.status}`,
              err,
              responseValue,
              void 0,
              void 0,
              void 0
            );
          }
        }
        default: {
          try {
            const responseValue = await response.body("json");
            throw new import_shared_errors3.ClientRequestError(
              `${response.status}`,
              response,
              responseValue,
              getErrorMessage(
                responseValue,
                `Unexpected api response: code=${response.status} body="${responseValue}"`
              )
            );
          } catch (error) {
            if (error instanceof import_shared_errors3.ClientRequestError) {
              throw error;
            }
            const responseValue = await response.body("text").catch(() => "");
            throw new import_shared_errors3.InvalidResponseError(
              `${response.status}`,
              error,
              responseValue,
              "json",
              "Unexpected api response",
              response
            );
          }
        }
      }
    } catch (e) {
      if ((0, import_http_client.isHttpClientAbortError)(e) || (0, import_shared_errors3.isInvalidResponseError)(e) || (0, import_shared_errors3.isClientRequestError)(e)) {
        throw e;
      }
      if (!(0, import_http_client.isHttpClientResponseError)(e)) {
        throw new import_shared_errors3.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AnalyzerErrorEnvelopeError(
              `400`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Invalid or malformed request."),
              e
            );
          } catch (err) {
            if (err instanceof AnalyzerErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `AnalyzersClient.executeAnalyzer:400`,
              err,
              responseValue,
              "AnalyzerErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AnalyzerErrorEnvelopeError(
              `403`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Insufficient permissions."),
              e
            );
          } catch (err) {
            if (err instanceof AnalyzerErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `AnalyzersClient.executeAnalyzer:403`,
              err,
              responseValue,
              "AnalyzerErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AnalyzerErrorEnvelopeError(
              `404`,
              response,
              errorBody,
              getErrorMessage(errorBody, "No analyzer available with the requested name."),
              e
            );
          } catch (err) {
            if (err instanceof AnalyzerErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `AnalyzersClient.executeAnalyzer:404`,
              err,
              responseValue,
              "AnalyzerErrorEnvelope",
              void 0,
              response
            );
          }
        }
        default: {
          if (response.status >= 400 && response.status <= 499) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
              throw new AnalyzerErrorEnvelopeError(
                `${response.status}`,
                response,
                errorBody,
                getErrorMessage(errorBody, "Client error."),
                e
              );
            } catch (err) {
              if (err instanceof AnalyzerErrorEnvelopeError) {
                throw err;
              }
              throw new import_shared_errors3.InvalidResponseError(
                `AnalyzersClient.executeAnalyzer:${response.status}`,
                err,
                responseValue,
                "AnalyzerErrorEnvelope",
                void 0,
                response
              );
            }
          } else if (response.status >= 500 && response.status <= 599) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
              throw new AnalyzerErrorEnvelopeError(
                `${response.status}`,
                response,
                errorBody,
                getErrorMessage(errorBody, "Internal server error."),
                e
              );
            } catch (err) {
              if (err instanceof AnalyzerErrorEnvelopeError) {
                throw err;
              }
              throw new import_shared_errors3.InvalidResponseError(
                `AnalyzersClient.executeAnalyzer:${response.status}`,
                err,
                responseValue,
                "AnalyzerErrorEnvelope",
                void 0,
                response
              );
            }
          } else {
            try {
              const responseValue = await response.body("json");
              throw new import_shared_errors3.ClientRequestError(
                `${response.status}`,
                response,
                responseValue,
                getErrorMessage(
                  responseValue,
                  `Unexpected api response: code=${response.status} body="${responseValue}"`
                )
              );
            } catch (error) {
              if (error instanceof import_shared_errors3.ClientRequestError) {
                throw error;
              }
              const responseValue = await response.body("text").catch(() => "");
              throw new import_shared_errors3.InvalidResponseError(
                `${response.status}`,
                error,
                responseValue,
                "json",
                "Unexpected api response",
                response
              );
            }
          }
        }
      }
    }
  }
  /**
   * Poll for the result of a started analyzer execution.
   *
   * **Required scope:** davis:analyzers:execute<br/><br/>Use the request token returned when starting an analyzer execution to poll for a result.
   *
   * The `executionStatus` of the result indicates if the analyzer execution is already completed.
   * While the status is `RUNNING`, wait and poll for the result again in subsequent calls.
   *
   * @returns {Promise<AnalyzerPollResult>} The current analyzer execution status and an optional partial result.
   *
   * @example <caption>Code example</caption>
   * import { analyzersClient } from "@dynatrace-sdk/client-davis-analyzers";
   *
   * const data = await analyzersClient.pollAnalyzerExecution({
   *   analyzerName: "...",
   *   requestToken: "RHluYXRyYWNlMjAyMw==",
   * });
   *
   * @throws {AnalyzerErrorEnvelopeError} Invalid or malformed request. | Insufficient permissions. | No analyzer available with the requested name. | The analyzer result with the request token has already been consumed or its time-to-live (TTL) has been reached. The result is not available anymore. | Client error. | Internal server error.
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async pollAnalyzerExecution(config) {
    if (!config) {
      throw new import_shared_errors3.ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const query = toQueryString({ "request-token": config.requestToken, "timeout-seconds": config.timeoutSeconds });
    try {
      const response = await this.httpClient.send({
        url: `/platform/davis/analyzers/v1/analyzers/${config.analyzerName}:poll${query}`,
        method: "GET",
        headers: {
          Accept: "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
      } catch (err) {
        throw new import_shared_errors3.InvalidResponseError(
          `AnalyzersClient.pollAnalyzerExecution:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if ((0, import_http_client.isHttpClientAbortError)(e) || (0, import_shared_errors3.isInvalidResponseError)(e)) {
        throw e;
      }
      if (!(0, import_http_client.isHttpClientResponseError)(e)) {
        throw new import_shared_errors3.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AnalyzerErrorEnvelopeError(
              `400`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Invalid or malformed request."),
              e
            );
          } catch (err) {
            if (err instanceof AnalyzerErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `AnalyzersClient.pollAnalyzerExecution:400`,
              err,
              responseValue,
              "AnalyzerErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AnalyzerErrorEnvelopeError(
              `403`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Insufficient permissions."),
              e
            );
          } catch (err) {
            if (err instanceof AnalyzerErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `AnalyzersClient.pollAnalyzerExecution:403`,
              err,
              responseValue,
              "AnalyzerErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AnalyzerErrorEnvelopeError(
              `404`,
              response,
              errorBody,
              getErrorMessage(errorBody, "No analyzer available with the requested name."),
              e
            );
          } catch (err) {
            if (err instanceof AnalyzerErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `AnalyzersClient.pollAnalyzerExecution:404`,
              err,
              responseValue,
              "AnalyzerErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 410: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AnalyzerErrorEnvelopeError(
              `410`,
              response,
              errorBody,
              getErrorMessage(
                errorBody,
                "The analyzer result with the request token has already been consumed or its time-to-live (TTL) has been reached. The result is not available anymore."
              ),
              e
            );
          } catch (err) {
            if (err instanceof AnalyzerErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `AnalyzersClient.pollAnalyzerExecution:410`,
              err,
              responseValue,
              "AnalyzerErrorEnvelope",
              void 0,
              response
            );
          }
        }
        default: {
          if (response.status >= 400 && response.status <= 499) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
              throw new AnalyzerErrorEnvelopeError(
                `${response.status}`,
                response,
                errorBody,
                getErrorMessage(errorBody, "Client error."),
                e
              );
            } catch (err) {
              if (err instanceof AnalyzerErrorEnvelopeError) {
                throw err;
              }
              throw new import_shared_errors3.InvalidResponseError(
                `AnalyzersClient.pollAnalyzerExecution:${response.status}`,
                err,
                responseValue,
                "AnalyzerErrorEnvelope",
                void 0,
                response
              );
            }
          } else if (response.status >= 500 && response.status <= 599) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
              throw new AnalyzerErrorEnvelopeError(
                `${response.status}`,
                response,
                errorBody,
                getErrorMessage(errorBody, "Internal server error."),
                e
              );
            } catch (err) {
              if (err instanceof AnalyzerErrorEnvelopeError) {
                throw err;
              }
              throw new import_shared_errors3.InvalidResponseError(
                `AnalyzersClient.pollAnalyzerExecution:${response.status}`,
                err,
                responseValue,
                "AnalyzerErrorEnvelope",
                void 0,
                response
              );
            }
          } else {
            try {
              const responseValue = await response.body("json");
              throw new import_shared_errors3.ClientRequestError(
                `${response.status}`,
                response,
                responseValue,
                getErrorMessage(
                  responseValue,
                  `Unexpected api response: code=${response.status} body="${responseValue}"`
                )
              );
            } catch (error) {
              if (error instanceof import_shared_errors3.ClientRequestError) {
                throw error;
              }
              const responseValue = await response.body("text").catch(() => "");
              throw new import_shared_errors3.InvalidResponseError(
                `${response.status}`,
                error,
                responseValue,
                "json",
                "Unexpected api response",
                response
              );
            }
          }
        }
      }
    }
  }
  /**
   * Stop a started analyzer execution.
   *
   * **Required scope:** davis:analyzers:execute<br/><br/>Use the request token that was returned when starting an analyzer execution to cancel it.
   *
   * @returns {Promise<AnalyzerCancelResult | void>} The current analyzer execution status and an optional partial result.
   *
   * @example <caption>Code example</caption>
   * import { analyzersClient } from "@dynatrace-sdk/client-davis-analyzers";
   *
   * const data = await analyzersClient.cancelAnalyzerExecution({
   *   analyzerName: "...",
   *   requestToken: "RHluYXRyYWNlMjAyMw==",
   * });
   *
   * @throws {AnalyzerErrorEnvelopeError} Invalid or malformed request. | Insufficient permissions. | No analyzer available with the requested name. | The analyzer result with the request token has already been consumed or its time-to-live (TTL) has been reached. The result is not available anymore. | Client error. | Internal server error.
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async cancelAnalyzerExecution(config) {
    if (!config) {
      throw new import_shared_errors3.ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const query = toQueryString({ "request-token": config.requestToken });
    try {
      const response = await this.httpClient.send({
        url: `/platform/davis/analyzers/v1/analyzers/${config.analyzerName}:cancel${query}`,
        method: "POST",
        headers: {
          Accept: "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200, 202].includes(status);
        }
      });
      switch (response.status) {
        case 200: {
          const responseValue = await response.body("json");
          try {
            return this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
          } catch (err) {
            throw new import_shared_errors3.InvalidResponseError(
              `AnalyzersClient.cancelAnalyzerExecution:${response.status}`,
              err,
              responseValue,
              void 0,
              void 0,
              void 0
            );
          }
        }
        case 202: {
          return;
        }
        default: {
          try {
            const responseValue = await response.body("json");
            throw new import_shared_errors3.ClientRequestError(
              `${response.status}`,
              response,
              responseValue,
              getErrorMessage(
                responseValue,
                `Unexpected api response: code=${response.status} body="${responseValue}"`
              )
            );
          } catch (error) {
            if (error instanceof import_shared_errors3.ClientRequestError) {
              throw error;
            }
            const responseValue = await response.body("text").catch(() => "");
            throw new import_shared_errors3.InvalidResponseError(
              `${response.status}`,
              error,
              responseValue,
              "json",
              "Unexpected api response",
              response
            );
          }
        }
      }
    } catch (e) {
      if ((0, import_http_client.isHttpClientAbortError)(e) || (0, import_shared_errors3.isInvalidResponseError)(e) || (0, import_shared_errors3.isClientRequestError)(e)) {
        throw e;
      }
      if (!(0, import_http_client.isHttpClientResponseError)(e)) {
        throw new import_shared_errors3.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AnalyzerErrorEnvelopeError(
              `400`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Invalid or malformed request."),
              e
            );
          } catch (err) {
            if (err instanceof AnalyzerErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `AnalyzersClient.cancelAnalyzerExecution:400`,
              err,
              responseValue,
              "AnalyzerErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AnalyzerErrorEnvelopeError(
              `403`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Insufficient permissions."),
              e
            );
          } catch (err) {
            if (err instanceof AnalyzerErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `AnalyzersClient.cancelAnalyzerExecution:403`,
              err,
              responseValue,
              "AnalyzerErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AnalyzerErrorEnvelopeError(
              `404`,
              response,
              errorBody,
              getErrorMessage(errorBody, "No analyzer available with the requested name."),
              e
            );
          } catch (err) {
            if (err instanceof AnalyzerErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `AnalyzersClient.cancelAnalyzerExecution:404`,
              err,
              responseValue,
              "AnalyzerErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 410: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AnalyzerErrorEnvelopeError(
              `410`,
              response,
              errorBody,
              getErrorMessage(
                errorBody,
                "The analyzer result with the request token has already been consumed or its time-to-live (TTL) has been reached. The result is not available anymore."
              ),
              e
            );
          } catch (err) {
            if (err instanceof AnalyzerErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `AnalyzersClient.cancelAnalyzerExecution:410`,
              err,
              responseValue,
              "AnalyzerErrorEnvelope",
              void 0,
              response
            );
          }
        }
        default: {
          if (response.status >= 400 && response.status <= 499) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
              throw new AnalyzerErrorEnvelopeError(
                `${response.status}`,
                response,
                errorBody,
                getErrorMessage(errorBody, "Client error."),
                e
              );
            } catch (err) {
              if (err instanceof AnalyzerErrorEnvelopeError) {
                throw err;
              }
              throw new import_shared_errors3.InvalidResponseError(
                `AnalyzersClient.cancelAnalyzerExecution:${response.status}`,
                err,
                responseValue,
                "AnalyzerErrorEnvelope",
                void 0,
                response
              );
            }
          } else if (response.status >= 500 && response.status <= 599) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
              throw new AnalyzerErrorEnvelopeError(
                `${response.status}`,
                response,
                errorBody,
                getErrorMessage(errorBody, "Internal server error."),
                e
              );
            } catch (err) {
              if (err instanceof AnalyzerErrorEnvelopeError) {
                throw err;
              }
              throw new import_shared_errors3.InvalidResponseError(
                `AnalyzersClient.cancelAnalyzerExecution:${response.status}`,
                err,
                responseValue,
                "AnalyzerErrorEnvelope",
                void 0,
                response
              );
            }
          } else {
            try {
              const responseValue = await response.body("json");
              throw new import_shared_errors3.ClientRequestError(
                `${response.status}`,
                response,
                responseValue,
                getErrorMessage(
                  responseValue,
                  `Unexpected api response: code=${response.status} body="${responseValue}"`
                )
              );
            } catch (error) {
              if (error instanceof import_shared_errors3.ClientRequestError) {
                throw error;
              }
              const responseValue = await response.body("text").catch(() => "");
              throw new import_shared_errors3.InvalidResponseError(
                `${response.status}`,
                error,
                responseValue,
                "json",
                "Unexpected api response",
                response
              );
            }
          }
        }
      }
    }
  }
  /**
   * Validate the provided input for an analyzer execution.
   *
   * **Required scope:** davis:analyzers:execute<br/><br/>Provide the same input that is planned to be used for an analyzer execution to have it validated without
   * executing the analyzer.
   *
   * @returns {Promise<AnalyzerValidationResult>} The validation result for the requested analyzer execution.
   *
   * @example <caption>Code example</caption>
   * import { analyzersClient } from "@dynatrace-sdk/client-davis-analyzers";
   *
   * const data =
   *   await analyzersClient.validateAnalyzerExecution({
   *     analyzerName: "...",
   *     body: {},
   *   });
   *
   * @throws {AnalyzerErrorEnvelopeError} No analyzer available with the requested name. | Insufficient permissions. | Not found | Client error. | Internal server error.
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async validateAnalyzerExecution(config) {
    if (!config) {
      throw new import_shared_errors3.ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const encodedBody = this.shouldTransformDates ? transformRequest(config.body, []) : config.body;
    try {
      const response = await this.httpClient.send({
        url: `/platform/davis/analyzers/v1/analyzers/${config.analyzerName}:validate`,
        method: "POST",
        requestBodyType: "json",
        body: encodedBody,
        headers: {
          "Content-Type": "application/json",
          Accept: "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
      } catch (err) {
        throw new import_shared_errors3.InvalidResponseError(
          `AnalyzersClient.validateAnalyzerExecution:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if ((0, import_http_client.isHttpClientAbortError)(e) || (0, import_shared_errors3.isInvalidResponseError)(e)) {
        throw e;
      }
      if (!(0, import_http_client.isHttpClientResponseError)(e)) {
        throw new import_shared_errors3.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AnalyzerErrorEnvelopeError(
              `400`,
              response,
              errorBody,
              getErrorMessage(errorBody, "No analyzer available with the requested name."),
              e
            );
          } catch (err) {
            if (err instanceof AnalyzerErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `AnalyzersClient.validateAnalyzerExecution:400`,
              err,
              responseValue,
              "AnalyzerErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AnalyzerErrorEnvelopeError(
              `403`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Insufficient permissions."),
              e
            );
          } catch (err) {
            if (err instanceof AnalyzerErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `AnalyzersClient.validateAnalyzerExecution:403`,
              err,
              responseValue,
              "AnalyzerErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AnalyzerErrorEnvelopeError(
              `404`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Not found"),
              e
            );
          } catch (err) {
            if (err instanceof AnalyzerErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `AnalyzersClient.validateAnalyzerExecution:404`,
              err,
              responseValue,
              "AnalyzerErrorEnvelope",
              void 0,
              response
            );
          }
        }
        default: {
          if (response.status >= 400 && response.status <= 499) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
              throw new AnalyzerErrorEnvelopeError(
                `${response.status}`,
                response,
                errorBody,
                getErrorMessage(errorBody, "Client error."),
                e
              );
            } catch (err) {
              if (err instanceof AnalyzerErrorEnvelopeError) {
                throw err;
              }
              throw new import_shared_errors3.InvalidResponseError(
                `AnalyzersClient.validateAnalyzerExecution:${response.status}`,
                err,
                responseValue,
                "AnalyzerErrorEnvelope",
                void 0,
                response
              );
            }
          } else if (response.status >= 500 && response.status <= 599) {
            const responseValue = await response.body("json");
            try {
              const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
              throw new AnalyzerErrorEnvelopeError(
                `${response.status}`,
                response,
                errorBody,
                getErrorMessage(errorBody, "Internal server error."),
                e
              );
            } catch (err) {
              if (err instanceof AnalyzerErrorEnvelopeError) {
                throw err;
              }
              throw new import_shared_errors3.InvalidResponseError(
                `AnalyzersClient.validateAnalyzerExecution:${response.status}`,
                err,
                responseValue,
                "AnalyzerErrorEnvelope",
                void 0,
                response
              );
            }
          } else {
            try {
              const responseValue = await response.body("json");
              throw new import_shared_errors3.ClientRequestError(
                `${response.status}`,
                response,
                responseValue,
                getErrorMessage(
                  responseValue,
                  `Unexpected api response: code=${response.status} body="${responseValue}"`
                )
              );
            } catch (error) {
              if (error instanceof import_shared_errors3.ClientRequestError) {
                throw error;
              }
              const responseValue = await response.body("text").catch(() => "");
              throw new import_shared_errors3.InvalidResponseError(
                `${response.status}`,
                error,
                responseValue,
                "json",
                "Unexpected api response",
                response
              );
            }
          }
        }
      }
    }
  }
};
var analyzersClient = /* @__PURE__ */ new AnalyzersClient(import_http_client.httpClient);

// packages/client/davis-analyzers/src/lib/error-envelopes/index.ts
var import_shared_errors4 = require("@dynatrace-sdk/shared-errors");

// packages/client/davis-analyzers/src/lib/models/analyzer-definition-type.ts
var AnalyzerDefinitionType = /* @__PURE__ */ ((AnalyzerDefinitionType2) => {
  AnalyzerDefinitionType2["Implemented"] = "IMPLEMENTED";
  AnalyzerDefinitionType2["Interface"] = "INTERFACE";
  return AnalyzerDefinitionType2;
})(AnalyzerDefinitionType || {});

// packages/client/davis-analyzers/src/lib/models/analyzer-dimensional-data-type.ts
var AnalyzerDimensionalDataType = /* @__PURE__ */ ((AnalyzerDimensionalDataType2) => {
  AnalyzerDimensionalDataType2["EntityId"] = "entityId";
  AnalyzerDimensionalDataType2["TimeseriesArray"] = "timeseriesArray";
  return AnalyzerDimensionalDataType2;
})(AnalyzerDimensionalDataType || {});

// packages/client/davis-analyzers/src/lib/models/analyzer-execution-log-level.ts
var AnalyzerExecutionLogLevel = /* @__PURE__ */ ((AnalyzerExecutionLogLevel2) => {
  AnalyzerExecutionLogLevel2["Tracing"] = "TRACING";
  AnalyzerExecutionLogLevel2["Info"] = "INFO";
  AnalyzerExecutionLogLevel2["Warning"] = "WARNING";
  AnalyzerExecutionLogLevel2["Severe"] = "SEVERE";
  return AnalyzerExecutionLogLevel2;
})(AnalyzerExecutionLogLevel || {});

// packages/client/davis-analyzers/src/lib/models/analyzer-output-log-level.ts
var AnalyzerOutputLogLevel = /* @__PURE__ */ ((AnalyzerOutputLogLevel2) => {
  AnalyzerOutputLogLevel2["Tracing"] = "TRACING";
  AnalyzerOutputLogLevel2["Info"] = "INFO";
  AnalyzerOutputLogLevel2["Warning"] = "WARNING";
  return AnalyzerOutputLogLevel2;
})(AnalyzerOutputLogLevel || {});

// packages/client/davis-analyzers/src/lib/models/analyzer-result-execution-status.ts
var AnalyzerResultExecutionStatus = /* @__PURE__ */ ((AnalyzerResultExecutionStatus2) => {
  AnalyzerResultExecutionStatus2["Running"] = "RUNNING";
  AnalyzerResultExecutionStatus2["Aborted"] = "ABORTED";
  AnalyzerResultExecutionStatus2["Completed"] = "COMPLETED";
  return AnalyzerResultExecutionStatus2;
})(AnalyzerResultExecutionStatus || {});

// packages/client/davis-analyzers/src/lib/models/analyzer-result-result-status.ts
var AnalyzerResultResultStatus = /* @__PURE__ */ ((AnalyzerResultResultStatus2) => {
  AnalyzerResultResultStatus2["Successful"] = "SUCCESSFUL";
  AnalyzerResultResultStatus2["SuccessfulWithWarnings"] = "SUCCESSFUL_WITH_WARNINGS";
  AnalyzerResultResultStatus2["Failed"] = "FAILED";
  return AnalyzerResultResultStatus2;
})(AnalyzerResultResultStatus || {});

// packages/client/davis-analyzers/src/lib/models/json-schema-config.ts
var JsonSchemaConfig = /* @__PURE__ */ ((JsonSchemaConfig2) => {
  JsonSchemaConfig2["Default"] = "DEFAULT";
  JsonSchemaConfig2["Llm"] = "LLM";
  return JsonSchemaConfig2;
})(JsonSchemaConfig || {});

// packages/client/davis-analyzers/src/lib/models/parameter-definition-type.ts
var ParameterDefinitionType = /* @__PURE__ */ ((ParameterDefinitionType2) => {
  ParameterDefinitionType2["AnalyzerStructure"] = "analyzerStructure";
  ParameterDefinitionType2["Boolean"] = "boolean";
  ParameterDefinitionType2["DimensionQuery"] = "dimensionQuery";
  ParameterDefinitionType2["Enumeration"] = "enumeration";
  ParameterDefinitionType2["Integer"] = "integer";
  ParameterDefinitionType2["Number"] = "number";
  ParameterDefinitionType2["String"] = "string";
  ParameterDefinitionType2["Timeframe"] = "timeframe";
  ParameterDefinitionType2["TimeseriesArray"] = "timeseriesArray";
  ParameterDefinitionType2["TimeseriesQuery"] = "timeseriesQuery";
  ParameterDefinitionType2["Timestamp"] = "timestamp";
  return ParameterDefinitionType2;
})(ParameterDefinitionType || {});
