// @dynatrace/strato-components/esm/filters/segment-selector/parts/EmptySegments.js
import { jsx, jsxs } from "react/jsx-runtime";
import { forwardRef, useContext, useEffect, useRef } from "react";
import { useIntl } from "react-intl";
import { ContainerIcon, ExternalLinkIcon } from "@dynatrace/strato-icons";
import { Button } from "../../../buttons/button/Button.js";
import { SEGMENTS_DOCUMENTATION_URL } from "../../../core/external-links.js";
import { Flex } from "../../../layouts/flex/Flex.js";
import { Heading } from "../../../typography/heading/Heading.js";
import { segmentsClient } from "../contexts/SegmentsCache/SegmentsCache.js";
import { UISegmentsContext } from "../contexts/SegmentsUI/SegmentsUIContext.js";
import { SEGMENTSELECTOR_MESSAGES } from "../segmentMessages.js";
import { AppLinkButtonCSS } from "./EmptySegments.sty.js";
import { NoAccessWithoutSelection } from "./noAccessRights/NoAccessWithoutSelection.js";
import * as styles from "./Overlay.sty.js";
import { getSegmentManagementHref } from "./segmentManagementHref.js";
const EmptySegments = forwardRef(
  (props, forwardedRef) => {
    const intl = useIntl();
    const buttonRef = useRef(null);
    const { readOnly } = useContext(UISegmentsContext);
    useEffect(() => {
      buttonRef.current?.focus();
    }, []);
    if (segmentsClient.hasAccessRights === false) {
      return /* @__PURE__ */ jsx(NoAccessWithoutSelection, {});
    }
    return /* @__PURE__ */ jsxs(
      Flex,
      {
        flexDirection: "column",
        alignItems: "center",
        padding: 16,
        ref: forwardedRef,
        width: "100vw",
        className: styles.overlayDefaultWidth,
        style: { maxWidth: "320px" },
        children: [
          /* @__PURE__ */ jsx(ContainerIcon, {}),
          /* @__PURE__ */ jsx(Heading, { level: 6, children: intl.formatMessage(SEGMENTSELECTOR_MESSAGES.noSegmentsConfigured) }),
          /* @__PURE__ */ jsxs(
            Flex,
            {
              paddingTop: 8,
              width: "100%",
              justifyContent: readOnly ? "center" : "space-between",
              children: [
                !readOnly && /* @__PURE__ */ jsx(
                  Button,
                  {
                    as: "a",
                    ref: buttonRef,
                    href: getSegmentManagementHref(),
                    target: "_blank",
                    rel: "noopener noreferrer",
                    variant: "emphasized",
                    className: AppLinkButtonCSS,
                    children: intl.formatMessage(SEGMENTSELECTOR_MESSAGES.manageSegments)
                  }
                ),
                /* @__PURE__ */ jsxs(
                  Button,
                  {
                    as: "a",
                    href: SEGMENTS_DOCUMENTATION_URL,
                    target: "_blank",
                    rel: "noopener noreferrer",
                    children: [
                      intl.formatMessage(SEGMENTSELECTOR_MESSAGES.docsLinkText),
                      /* @__PURE__ */ jsx(Button.Suffix, { children: /* @__PURE__ */ jsx(ExternalLinkIcon, {}) })
                    ]
                  }
                )
              ]
            }
          )
        ]
      }
    );
  }
);
export {
  EmptySegments
};
