// @dynatrace/strato-components/esm/filters/segment-selector/parts/LoadingSegments.js
import { jsx, jsxs } from "react/jsx-runtime";
import { forwardRef } from "react";
import { useIntl } from "react-intl";
import { Skeleton } from "../../../content/skeleton/Skeleton.js";
import { Flex } from "../../../layouts/flex/Flex.js";
import { SEGMENTSELECTOR_MESSAGES } from "../segmentMessages.js";
import * as styles from "./Overlay.sty.js";
const LoadingSegments = forwardRef(
  (props, forwardedRef) => {
    const intl = useIntl();
    return /* @__PURE__ */ jsxs(
      Flex,
      {
        flexDirection: "column",
        alignItems: "center",
        padding: 16,
        ref: forwardedRef,
        width: "100vw",
        className: styles.overlayDefaultWidth,
        "aria-label": intl.formatMessage(
          SEGMENTSELECTOR_MESSAGES.loadingAriaLabel
        ),
        children: [
          /* @__PURE__ */ jsx(Skeleton, { width: "100%", height: "20px" }),
          /* @__PURE__ */ jsx(Skeleton, { width: "100%", height: "20px" }),
          /* @__PURE__ */ jsx(Skeleton, { width: "100%", height: "20px" })
        ]
      }
    );
  }
);
export {
  LoadingSegments
};
