// @dynatrace/strato-components/esm/filters/segment-selector/parts/Overlay.js
import { jsx, jsxs } from "react/jsx-runtime";
import { useQuery } from "@tanstack/react-query";
import {
  Fragment,
  useContext,
  useEffect,
  useMemo,
  useRef
} from "react";
import { FormattedMessage, useIntl } from "react-intl";
import { PlusIcon } from "@dynatrace/strato-icons";
import { getSegmentsPresets } from "@dynatrace-sdk/segments-presets";
import { EmptySegments } from "./EmptySegments.js";
import { LoadingSegments } from "./LoadingSegments.js";
import { NoAccessHint } from "./noAccessRights/NoAccessHint.js";
import * as overlayStyles from "./Overlay.sty.js";
import { SegmentFooter } from "./SegmentFooter.js";
import { SegmentsPresetList } from "./segmentsPresets/SegmentsPresetsList.js";
import { SelectionBar } from "./selectionBar/SelectionBar.js";
import { Button } from "../../../buttons/button/Button.js";
import { useId } from "../../../core/hooks/useId.js";
import { FieldSet } from "../../../forms/forms-core/components/field-set/FieldSet.js";
import { FormFieldMessageContextProvider } from "../../../forms/forms-core/components/form-field/FormFieldMessageContextProvider.js";
import { useFormFieldMessagesContext } from "../../../forms/forms-core/hooks/useFormFieldMessagesContext.js";
import { FormFieldMessagesAction } from "../../../forms/forms-core/state/formfield-messages-state-reducer.js";
import { Divider } from "../../../layouts/divider/Divider.js";
import { Flex } from "../../../layouts/flex/Flex.js";
import { Grid } from "../../../layouts/grid/Grid.js";
import {
  segmentsClient,
  useSegmentsCache
} from "../contexts/SegmentsCache/SegmentsCache.js";
import {
  UISegmentsContext,
  useSegmentsUIContext
} from "../contexts/SegmentsUI/SegmentsUIContext.js";
import { UISegmentsDispatchContext } from "../contexts/SegmentsUI/SegmentsUIDispatchContext.js";
import {
  areUiSegmentsValid,
  getErrorKey
} from "../contexts/SegmentsUI/segmentValidationHelper.js";
import { useSegments } from "../hooks/useSegments.js";
import { SEGMENTSELECTOR_MESSAGES } from "../segmentMessages.js";
import * as styles from "../SegmentSelector.sty.js";
const SegmentsOverlay = ({ labelId }) => {
  const { getAllSegments } = useSegmentsCache();
  const { data, isLoading } = useQuery(
    {
      queryKey: ["getAllSegments"],
      queryFn: () => getAllSegments().catch(() => /* @__PURE__ */ new Map()),
      notifyOnChangeProps: ["dataUpdatedAt"],
      throwOnError: false
    },
    segmentsClient.queryClient
  );
  const { isLoading: isLoadingPresets } = useQuery(
    {
      queryKey: ["getSegmentsPresets"],
      queryFn: () => (
        // We don't want to render the error fallback in case the presets are not available.
        // However, we sill want to keep the loading state until both queries are resolved.
        getSegmentsPresets().catch(() => {
          return {
            pinnedPresets: [],
            recentPresets: []
          };
        })
      )
    },
    segmentsClient.queryClient
  );
  if (isLoading || isLoadingPresets) {
    return /* @__PURE__ */ jsx(LoadingSegments, {});
  }
  return /* @__PURE__ */ jsx(SegmentsContent, { cachedSegments: data ?? /* @__PURE__ */ new Map(), id: labelId });
};
async function applySegments(uiSegments, setSegments, overlayState) {
  try {
    const applicableSegments = await Promise.all(
      uiSegments.filter((uiSegment) => !uiSegment.isPlaceholder).map(async (uiSegment) => {
        const errorKey = await getErrorKey(uiSegment);
        if (errorKey === "needs-update") {
          return {
            id: uiSegment.id
          };
        }
        return {
          id: uiSegment.id,
          variables: uiSegment.variables
        };
      })
    );
    await setSegments(applicableSegments);
    const isValid = await areUiSegmentsValid(applicableSegments);
    if (isValid) {
      overlayState?.close();
    }
  } catch {
  }
}
const SegmentsContent = ({ cachedSegments, id }) => {
  const { dispatch, overlayState } = useContext(UISegmentsDispatchContext);
  const { dispatch: parentErrorDispatch } = useFormFieldMessagesContext();
  const intl = useIntl();
  const addButtonRef = useRef(null);
  const {
    uiSegments,
    hasSelection,
    readOnly,
    focusedSegmentId,
    arrayIdentifierErrorIds
  } = useContext(UISegmentsContext);
  const formId = useId();
  const formRef = useRef(null);
  useEffect(() => {
    const errorIds = arrayIdentifierErrorIds;
    return () => {
      if (!errorIds) {
        return;
      }
      if (errorIds.length > 0) {
        parentErrorDispatch?.({
          type: FormFieldMessagesAction.ADDMESSAGE,
          payload: {
            id: "array-identifier",
            validityStateKey: "customError",
            variant: "error",
            messageType: "formcontrol",
            message: /* @__PURE__ */ jsx(
              FormattedMessage,
              {
                ...SEGMENTSELECTOR_MESSAGES.arrayIdentifierError
              }
            )
          }
        });
        return;
      }
      parentErrorDispatch?.({
        type: FormFieldMessagesAction.REMOVEMESSAGE,
        payload: {
          id: "array-identifier",
          messageType: "formcontrol"
        }
      });
    };
  }, [arrayIdentifierErrorIds]);
  const availableSegments = useMemo(() => {
    const tmp = [];
    cachedSegments.forEach((value, key) => {
      if (!value.temporarilyApplied && !uiSegments.some((segment) => segment.id === key)) {
        tmp.push({ ...value.segment });
      }
    });
    return tmp.sort(
      (firstSegment, secondSegment) => firstSegment.name.localeCompare(secondSegment.name)
    );
  }, [cachedSegments, uiSegments]);
  const {
    segments,
    removeAllSegments: removeAllSegmentsFromStore,
    setSegments
  } = useSegments();
  const { removeAllUISegments, clearPlaceholder } = useSegmentsUIContext();
  if (segments.length === 0 && cachedSegments.size === 0) {
    return /* @__PURE__ */ jsx(EmptySegments, {});
  }
  const MessageWrapper = segmentsClient.hasAccessRights ? FormFieldMessageContextProvider : Fragment;
  return /* @__PURE__ */ jsxs(
    Flex,
    {
      px: 4,
      pb: 8,
      gap: 0,
      flexDirection: "column",
      id,
      width: "100vw",
      className: overlayStyles.overlayDefaultWidth,
      children: [
        /* @__PURE__ */ jsxs("div", { className: overlayStyles.scrollContainer, children: [
          /* @__PURE__ */ jsxs(
            Grid,
            {
              as: "form",
              ref: formRef,
              id: formId,
              px: 12,
              pt: 12,
              pb: 16,
              onInvalid: (event) => {
                event.stopPropagation();
                if (event.nativeEvent.target.validity.customError && !event.nativeEvent.target.validity.valueMissing) {
                  applySegments(uiSegments, setSegments, overlayState).catch(
                    () => void 0
                  );
                }
              },
              onSubmit: (event) => {
                event.preventDefault();
                event.stopPropagation();
                applySegments(uiSegments, setSegments, overlayState).catch(
                  () => void 0
                );
              },
              onKeyDown: (event) => {
                if (event.key === "Enter" && (event.ctrlKey || event.metaKey)) {
                  formRef?.current?.requestSubmit();
                }
              },
              children: [
                /* @__PURE__ */ jsxs(FieldSet, { className: overlayStyles.fieldSet, children: [
                  /* @__PURE__ */ jsx(FieldSet.Legend, { className: overlayStyles.legend, children: intl.formatMessage(SEGMENTSELECTOR_MESSAGES.legend) }),
                  /* @__PURE__ */ jsx("ul", { className: styles.segmentSelectorList, children: uiSegments?.map((segment) => /* @__PURE__ */ jsx(MessageWrapper, { children: /* @__PURE__ */ jsx(
                    SelectionBar,
                    {
                      uiSegment: segment,
                      segment: cachedSegments.get(segment.id),
                      readOnly,
                      availableSegments,
                      clearable: segment.isPlaceholder ? uiSegments.length > 1 : true,
                      autoFocus: segment.id === focusedSegmentId,
                      returnFocusTo: addButtonRef
                    },
                    segment.id
                  ) }, segment.id)) }),
                  !readOnly && hasSelection && segmentsClient.hasAccessRights !== false && /* @__PURE__ */ jsxs(Grid, { gridTemplateColumns: "1fr auto auto", children: [
                    /* @__PURE__ */ jsxs(
                      Button,
                      {
                        onClick: () => dispatch({
                          type: "add",
                          payload: { isPlaceholder: true }
                        }),
                        variant: "emphasized",
                        ref: addButtonRef,
                        "data-dt-name": "Add",
                        children: [
                          /* @__PURE__ */ jsx(
                            Button.Prefix,
                            {
                              "aria-label": intl.formatMessage(
                                SEGMENTSELECTOR_MESSAGES.addSegment
                              ),
                              children: /* @__PURE__ */ jsx(PlusIcon, {})
                            }
                          ),
                          "Segment"
                        ]
                      }
                    ),
                    /* @__PURE__ */ jsx(
                      Button,
                      {
                        onClick: () => {
                          removeAllSegmentsFromStore();
                          removeAllUISegments();
                          overlayState?.close();
                        },
                        type: "reset",
                        form: formId,
                        "data-dt-name": "ClearAll",
                        children: intl.formatMessage(SEGMENTSELECTOR_MESSAGES.clearAll)
                      }
                    ),
                    /* @__PURE__ */ jsx(
                      Button,
                      {
                        type: "submit",
                        form: formId,
                        color: "primary",
                        variant: "accent",
                        "data-dt-name": "Apply",
                        onClick: () => clearPlaceholder(),
                        children: intl.formatMessage(SEGMENTSELECTOR_MESSAGES.applyButton)
                      }
                    )
                  ] })
                ] }),
                segmentsClient.hasAccessRights === false && /* @__PURE__ */ jsx(NoAccessHint, { readOnly })
              ]
            }
          ),
          !readOnly && segmentsClient.hasAccessRights === true && /* @__PURE__ */ jsx(SegmentsPresetList, { showDivider: hasSelection })
        ] }),
        /* @__PURE__ */ jsx(Divider, {}),
        /* @__PURE__ */ jsx(SegmentFooter, {})
      ]
    }
  );
};
export {
  SegmentsOverlay
};
