// @dynatrace/strato-components/esm/filters/segment-selector/parts/SegmentFooter.js
import { jsx, jsxs } from "react/jsx-runtime";
import { useContext } from "react";
import { useIntl } from "react-intl";
import { SettingIcon } from "@dynatrace/strato-icons";
import { getSegmentManagementHref } from "./segmentManagementHref.js";
import { SEGMENTSELECTOR_MESSAGES } from "../segmentMessages.js";
import * as styles from "./Overlay.sty.js";
import { Button } from "../../../buttons/button/Button.js";
import { SEGMENTS_DOCUMENTATION_URL } from "../../../core/external-links.js";
import { useOverlayStacking } from "../../../core/hooks/useOverlayStacking.js";
import { Grid } from "../../../layouts/grid/Grid.js";
import { Tooltip } from "../../../overlays/tooltip/Tooltip.js";
import { ExternalLink } from "../../../typography/external-link/ExternalLink.js";
import { segmentsClient } from "../contexts/SegmentsCache/SegmentsCache.js";
import { UISegmentsContext } from "../contexts/SegmentsUI/SegmentsUIContext.js";
const SegmentFooter = () => {
  const intl = useIntl();
  const stackingContext = useOverlayStacking(void 0, "overlay");
  const { readOnly } = useContext(UISegmentsContext);
  return /* @__PURE__ */ jsxs(
    Grid,
    {
      rowGap: 8,
      columnGap: 0,
      px: 12,
      pb: segmentsClient.hasAccessRights ? 2 : 4,
      pt: 2,
      mt: 8,
      gridTemplateColumns: segmentsClient.hasAccessRights && !readOnly ? "auto 1fr" : "1fr",
      justifyItems: "end",
      alignItems: "center",
      children: [
        segmentsClient.hasAccessRights && !readOnly && /* @__PURE__ */ jsx(
          Tooltip,
          {
            placement: "right",
            text: intl.formatMessage(SEGMENTSELECTOR_MESSAGES.manageSegments),
            style: { zIndex: Number(stackingContext.style.zIndex) + 1 },
            children: /* @__PURE__ */ jsx(
              Button,
              {
                as: "a",
                href: getSegmentManagementHref(),
                target: "_blank",
                rel: "noopener noreferrer",
                size: "condensed",
                className: styles.link,
                "aria-label": intl.formatMessage(
                  SEGMENTSELECTOR_MESSAGES.manageSegments
                ),
                "data-dt-name": "ManageSegments",
                children: /* @__PURE__ */ jsx(SettingIcon, {})
              }
            )
          }
        ),
        /* @__PURE__ */ jsx(
          ExternalLink,
          {
            href: SEGMENTS_DOCUMENTATION_URL,
            className: styles.link,
            "data-dt-name": "LearnMoreAboutSegments",
            children: intl.formatMessage(SEGMENTSELECTOR_MESSAGES.docsLinkText)
          }
        )
      ]
    }
  );
};
export {
  SegmentFooter
};
