// @dynatrace/strato-components/esm/filters/segment-selector/parts/TooltipMessages.js
import { jsx, jsxs } from "react/jsx-runtime";
import { useSuspenseQueries } from "@tanstack/react-query";
import { Suspense } from "react";
import { useIntl, FormattedMessage } from "react-intl";
import { SkeletonText } from "../../../content/skeleton/SkeletonText.js";
import { Flex } from "../../../layouts/flex/Flex.js";
import { Text } from "../../../typography/text/Text.js";
import {
  segmentsClient,
  useSegmentsCache
} from "../contexts/SegmentsCache/SegmentsCache.js";
import { useSegments } from "../hooks/useSegments.js";
import { SEGMENTSELECTOR_MESSAGES } from "../segmentMessages.js";
import * as styles from "../SegmentSelector.sty.js";
import { getReadableValue } from "../utils.js";
const SegmentWithVariablesText = ({
  id,
  name,
  variables
}) => {
  const intl = useIntl();
  const endOfLine = variables && variables.length > 1 ? ";" : "";
  const values = variables?.map((value) => getReadableValue(value, intl)).join(", ") + endOfLine;
  return /* @__PURE__ */ jsxs("div", { style: { lineHeight: 1 }, children: [
    /* @__PURE__ */ jsxs(Text, { textStyle: "small-emphasized", children: [
      name,
      ": "
    ] }),
    /* @__PURE__ */ jsx(Text, { textStyle: "small", children: values })
  ] }, id);
};
const WrappedTooltipMessages = ({ ariaLabel }) => {
  const { segments } = useSegments();
  const { getSegment } = useSegmentsCache();
  const result = useSuspenseQueries(
    {
      queries: segments.map((segment) => ({
        queryKey: ["getSegment", segment.id],
        queryFn: () => getSegment(segment.id),
        select: (data) => ({
          selectedSegment: segment,
          fetchedSegment: data?.segment
        }),
        retryOnMount: false
      }))
    },
    segmentsClient.queryClient
  );
  if (segments.length === 0) {
    return /* @__PURE__ */ jsx(Text, { textStyle: "small-emphasized", className: styles.tooltipMaxHeight, children: ariaLabel ?? /* @__PURE__ */ jsx(FormattedMessage, { ...SEGMENTSELECTOR_MESSAGES.defaultAriaLabel }) });
  }
  return /* @__PURE__ */ jsx(Flex, { flexDirection: "column", gap: 4, className: styles.tooltipMaxHeight, children: result.map((query) => {
    const selectedSegment = query.data?.selectedSegment;
    const id = selectedSegment.id;
    const name = query.data?.fetchedSegment?.name ?? id;
    if (selectedSegment?.variables) {
      return /* @__PURE__ */ jsx(
        SegmentWithVariablesText,
        {
          id,
          name,
          variables: selectedSegment.variables?.[0]?.values
        },
        id
      );
    }
    return /* @__PURE__ */ jsx(Text, { textStyle: "small-emphasized", children: name }, id);
  }) });
};
const TooltipMessages = ({ ariaLabel }) => (
  /** Due to display: 'contents' on the tooltip we need to define a fixed width here. */
  /* @__PURE__ */ jsx(Suspense, { fallback: /* @__PURE__ */ jsx(SkeletonText, { width: "150px" }), children: /* @__PURE__ */ jsx(WrappedTooltipMessages, { ariaLabel }) })
);
export {
  TooltipMessages
};
