// @dynatrace/strato-components/charts/categorical-bar/hooks/useGenerateCategoricalBarData.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useGenerateCategoricalBarData_exports = {};
__export(useGenerateCategoricalBarData_exports, {
  useGenerateCategoricalBarData: () => useGenerateCategoricalBarData
});
module.exports = __toCommonJS(useGenerateCategoricalBarData_exports);
var import_useCategoricalBarChartBaseConfig = require("./useCategoricalBarChartBaseConfig.js");
var import_useColorRule = require("../../core/color-rule/useColorRule.js");
var import_ColorRuleContext = require("../../core/context/ColorRuleContext.js");
var import_color_palette = require("../../core/types/color-palette.js");
var import_is_valid_color_palette = require("../../core/utils/is-valid-color-palette.js");
var import_generate_categorical_bar_data_helpers = require("../utils/generate-categorical-bar-data-helpers.js");
var import_layout = require("../utils/layout.js");
function buildSeriesForColorRule(initialData) {
  const series = [];
  initialData.forEach((categoryData) => {
    if (typeof categoryData.value === "object" && categoryData.value !== null) {
      Object.entries(categoryData.value).forEach(
        ([dimensionKey, dimensionValue]) => {
          series.push({
            ...categoryData,
            value: { [dimensionKey]: dimensionValue },
            _originalData: categoryData,
            _dimensionKey: dimensionKey
          });
        }
      );
    } else {
      series.push({
        ...categoryData,
        _originalData: categoryData
      });
    }
  });
  return series;
}
function createCategoryMetadata(id, name) {
  return { id, name };
}
function useGenerateCategoricalBarData(initialData) {
  const {
    valueRepresentation,
    colorPalette: configuredColorPalette,
    colorPaletteMode,
    groupMode
  } = (0, import_useCategoricalBarChartBaseConfig.useCategoricalBarChartBaseConfig)();
  const isRelative = valueRepresentation === "relative";
  const firstCategoryUnit = initialData[0]?.unit;
  const colorRuleConfig = (0, import_ColorRuleContext.useColorRuleConfig)();
  const hasColorRules = colorRuleConfig.length > 0;
  const seriesForColorRule = hasColorRules ? buildSeriesForColorRule(initialData) : [];
  const colorRules = (0, import_useColorRule.useColorRule)(seriesForColorRule, "categoricalBar");
  const uniqueDimensions = /* @__PURE__ */ new Set();
  const categoricalData = [];
  const categoryMap = /* @__PURE__ */ new WeakMap();
  const dimensionMap = /* @__PURE__ */ new WeakMap();
  const colorByDimension = /* @__PURE__ */ new Map();
  const colorByCategory = /* @__PURE__ */ new Map();
  for (const categoryData of initialData) {
    const categoryId = crypto.randomUUID();
    const {
      unit,
      category: categoryName,
      value: categoryValue,
      color: categoryColor,
      ...rest
    } = categoryData;
    const newCategory = {
      id: categoryId,
      name: categoryName,
      unit,
      dimensions: [],
      ...rest
    };
    const isSingleDimension = unit !== firstCategoryUnit || typeof categoryValue !== "object";
    const categoricalColorPalette = colorPaletteMode === "multi-color" || (0, import_layout.isGrouped)(groupMode) || !isSingleDimension;
    const colorPalette = (0, import_is_valid_color_palette.isValidColorPalette)(configuredColorPalette) || (0, import_color_palette.isCustomColorPalette)(configuredColorPalette) ? configuredColorPalette : void 0;
    if (isSingleDimension) {
      const colorPositionIndex = initialData.indexOf(categoryData);
      const paletteColorPosition = categoricalColorPalette ? colorPositionIndex : 0;
      (0, import_generate_categorical_bar_data_helpers.handleSingleDimension)({
        categoryData,
        rest,
        categoryName,
        categoryColor,
        colorByDimension,
        colorRules,
        colorPalette,
        colorPaletteMode,
        paletteColorPosition,
        uniqueDimensions,
        dimensionMap,
        newCategory,
        seriesForColorRule,
        hasColorRules
      });
    } else {
      (0, import_generate_categorical_bar_data_helpers.handleMultipleDimensions)({
        categoryData,
        rest,
        categoryName,
        colorByDimension,
        colorByCategory,
        colorRules,
        colorPalette,
        categoricalColorPalette,
        uniqueDimensions,
        dimensionMap,
        newCategory,
        seriesForColorRule,
        hasColorRules
      });
    }
    const categoryMedatada = createCategoryMetadata(categoryId, categoryName);
    categoryMap.set(newCategory, categoryMedatada);
    categoricalData.push(newCategory);
  }
  return {
    categoryMap,
    dimensionMap,
    categoricalData,
    uniqueDimensions: Array.from(uniqueDimensions),
    unit: isRelative ? "percent" : firstCategoryUnit,
    valueRepresentation
  };
}
