// @dynatrace/strato-components/charts/categorical-bar/hooks/useGenerateFilteredCategoricalBarData.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useGenerateFilteredCategoricalBarData_exports = {};
__export(useGenerateFilteredCategoricalBarData_exports, {
  calculateRelativeValue: () => calculateRelativeValue,
  calculateTotalCategoryValue: () => calculateTotalCategoryValue,
  getMaxDimensionValue: () => getMaxDimensionValue,
  useGenerateFilteredCategoricalBarData: () => useGenerateFilteredCategoricalBarData
});
module.exports = __toCommonJS(useGenerateFilteredCategoricalBarData_exports);
var import_useCategoricalBarChartBaseConfig = require("./useCategoricalBarChartBaseConfig.js");
var import_get_relative_value = require("../../core/utils/get-relative-value.js");
var import_safe_get = require("../../core/utils/safe-get.js");
var import_layout = require("../utils/layout.js");
const getMaxDimensionValue = (filteredCategories) => Math.max(
  ...filteredCategories.map(
    (category) => category.dimensions.length > 0 ? Math.abs(category.dimensions[0].value) : 0
  )
);
const calculateTotalCategoryValue = (categoryData) => categoryData.dimensions.reduce(
  (total, dimension) => total + Math.abs(dimension.value),
  0
);
const calculateRelativeValue = (value, relativeTotalValue) => {
  const numericValue = Number(value);
  const numberValue = isNaN(numericValue) ? 0 : numericValue;
  return relativeTotalValue ? (0, import_get_relative_value.getRelativeValue)(numberValue, relativeTotalValue) : numberValue;
};
function useGenerateFilteredCategoricalBarData(filteredCategories, categoricalData, unit, categoryMap, dimensionMap, valueRepresentation) {
  const { groupMode } = (0, import_useCategoricalBarChartBaseConfig.useCategoricalBarChartBaseConfig)();
  const filteredUniqueDimensions = /* @__PURE__ */ new Set();
  const minMaxPerUnit = /* @__PURE__ */ new Map();
  let isSingleDimension = true;
  for (const filteredCategory of filteredCategories) {
    const { id: categoryId, dimensions: filteredDimensions } = filteredCategory;
    if (filteredDimensions.length > 1) {
      isSingleDimension = false;
    }
    const categoryObjReference = categoricalData.find(
      (p) => p.id === categoryId
    );
    const categoryMetadata = (0, import_safe_get.safeGet)(categoryMap, categoryObjReference);
    let relativeTotalValue = void 0;
    if (valueRepresentation === "relative") {
      relativeTotalValue = isSingleDimension ? getMaxDimensionValue(filteredCategories) : calculateTotalCategoryValue(filteredCategory);
    }
    const cumulativeRanges = [0, 0];
    for (const dimension of filteredDimensions) {
      const dimensionObjReference = categoryObjReference?.dimensions.find(
        (d) => d.name === dimension.name
      );
      const dimensionMetadata = (0, import_safe_get.safeGet)(dimensionMap, dimensionObjReference);
      let { value } = dimensionMetadata;
      if (valueRepresentation === "relative") {
        value = calculateRelativeValue(dimension.value, relativeTotalValue);
      }
      let cumulativeValue = 0;
      if ((0, import_layout.isStacked)(groupMode)) {
        const NEGATIVE_RANGE = 0;
        const POSITIVE_RANGE = 1;
        const rangeIndex = value < 0 ? NEGATIVE_RANGE : POSITIVE_RANGE;
        cumulativeValue = cumulativeRanges[rangeIndex];
        cumulativeRanges[rangeIndex] += value;
      }
      filteredUniqueDimensions.add(dimension.name);
      dimensionMetadata.value = value;
      dimensionMetadata.cumulativeValue = cumulativeValue;
      dimensionMap.set(dimension, dimensionMetadata);
      const valueToMinMax = (cumulativeValue ?? 0) + value;
      const currentMinMax = minMaxPerUnit.get(unit);
      if (!currentMinMax) {
        minMaxPerUnit.set(unit, {
          min: valueToMinMax,
          max: valueToMinMax
        });
      } else {
        minMaxPerUnit.set(unit, {
          min: Math.min(currentMinMax.min, valueToMinMax),
          max: Math.max(currentMinMax.max, valueToMinMax)
        });
      }
    }
    categoryMap.set(filteredCategory, categoryMetadata);
  }
  return {
    categoryMap,
    dimensionMap,
    filteredUniqueDimensions: Array.from(filteredUniqueDimensions),
    minMaxPerUnit,
    isSingleDimension
  };
}
