// @dynatrace/strato-components/esm/charts/core/types/color-palettes-definitions.js
import Colors from "@dynatrace/strato-design-tokens/colors";
import {
  apdexMap,
  logLevelMap,
  logLevelWithAlias,
  logStatusMap,
  logStatusWithAlias,
  securityRiskLevelMap,
  statusMap,
  thresholdMap,
  vulnerabilityStatusMap
} from "./color-palette.js";
import { prepareInvertedColorPalette } from "../utils/prepare-inverted-color-palette.js";
const colorPalettesWithAlias = prepareInvertedColorPalette({
  "log-status": logStatusWithAlias,
  "log-level": logLevelWithAlias
});
const themedColoringMap = prepareInvertedColorPalette({
  categorical: Colors.Charts.Categorical,
  "purple-rain": Colors.Charts.CategoricalThemed.PurpleRain,
  swamps: Colors.Charts.CategoricalThemed.Swamps,
  fireplace: Colors.Charts.CategoricalThemed.Fireplace,
  "blue-steel": Colors.Charts.CategoricalThemed.BlueSteel,
  rainbow: Colors.Charts.Rainbow
});
const internalAutomaticColoringMap = prepareInvertedColorPalette({
  apdex: Colors.Charts.Apdex,
  "security-risk-level": Colors.Charts.SecurityRiskLevel,
  "vulnerability-status": Colors.Charts.VulnerabilityStatus,
  threshold: Colors.Charts.Threshold,
  status: Colors.Charts.Status
});
const automaticColoringMap = prepareInvertedColorPalette({
  "log-status": logStatusMap,
  "log-level": logLevelMap,
  apdex: apdexMap,
  "security-risk-level": securityRiskLevelMap,
  "vulnerability-status": vulnerabilityStatusMap,
  threshold: thresholdMap,
  status: statusMap
});
const sequentialColoringMap = prepareInvertedColorPalette({
  "pink-purple": Colors.Charts.Sequential.PinkPurple,
  "blue-turquoise": Colors.Charts.Sequential.BlueTurquoise,
  blue: Colors.Charts.Sequential.Blue,
  grey: Colors.Charts.Sequential.Grey,
  moss: Colors.Charts.Sequential.Moss,
  turquoise: Colors.Charts.Sequential.Turquoise,
  pink: Colors.Charts.Sequential.Pink,
  yellow: Colors.Charts.Sequential.Yellow,
  petrol: Colors.Charts.Sequential.Petrol,
  purple: Colors.Charts.Sequential.Purple,
  pear: Colors.Charts.Sequential.Pear,
  brown: Colors.Charts.Sequential.Brown,
  red: Colors.Charts.Sequential.Red,
  magenta: Colors.Charts.Sequential.Magenta,
  orange: Colors.Charts.Sequential.Orange
});
const divergingColoringMap = prepareInvertedColorPalette({
  "orange-blue": Colors.Charts.Diverging.OrangeBlue,
  "red-blue": Colors.Charts.Diverging.RedBlue,
  "red-purple": Colors.Charts.Diverging.RedPurple,
  "blue-moss": Colors.Charts.Diverging.BlueMoss,
  "purple-yellow": Colors.Charts.Diverging.PurpleYellow,
  "red-green": Colors.Charts.Diverging.RedGreen,
  "orange-petrol": Colors.Charts.Diverging.OrangePetrol,
  "pink-green": Colors.Charts.Diverging.PinkGreen
});
const colorPaletteMap = {
  ...themedColoringMap,
  ...internalAutomaticColoringMap,
  ...colorPalettesWithAlias,
  ...sequentialColoringMap,
  ...divergingColoringMap
};
export {
  automaticColoringMap,
  colorPaletteMap,
  colorPalettesWithAlias,
  divergingColoringMap,
  sequentialColoringMap
};
