// @dynatrace/strato-components/esm/charts/core/types/color-palette.js
import Colors from "@dynatrace/strato-design-tokens/colors";
const themedColoringOptions = [
  "categorical",
  "purple-rain",
  "swamps",
  "fireplace",
  "blue-steel",
  "rainbow",
  "categorical-inverted",
  "purple-rain-inverted",
  "swamps-inverted",
  "fireplace-inverted",
  "blue-steel-inverted",
  "rainbow-inverted"
];
const automaticColoringOptions = [
  "log-status",
  "log-level",
  "apdex",
  "security-risk-level",
  "vulnerability-status",
  "threshold",
  "status",
  "log-status-inverted",
  "log-level-inverted",
  "apdex-inverted",
  "security-risk-level-inverted",
  "vulnerability-status-inverted",
  "threshold-inverted",
  "status-inverted"
];
const sequentialColoringOptions = [
  "pink-purple",
  "blue-turquoise",
  "blue",
  "grey",
  "moss",
  "turquoise",
  "pink",
  "yellow",
  "petrol",
  "purple",
  "pear",
  "brown",
  "red",
  "magenta",
  "orange",
  "pink-purple-inverted",
  "blue-turquoise-inverted",
  "blue-inverted",
  "grey-inverted",
  "moss-inverted",
  "turquoise-inverted",
  "pink-inverted",
  "yellow-inverted",
  "petrol-inverted",
  "purple-inverted",
  "pear-inverted",
  "brown-inverted",
  "red-inverted",
  "magenta-inverted",
  "orange-inverted"
];
const divergingColoringOptions = [
  "orange-blue",
  "red-blue",
  "red-purple",
  "blue-moss",
  "purple-yellow",
  "red-green",
  "orange-petrol",
  "pink-green",
  "orange-blue-inverted",
  "red-blue-inverted",
  "red-purple-inverted",
  "blue-moss-inverted",
  "purple-yellow-inverted",
  "red-green-inverted",
  "orange-petrol-inverted",
  "pink-green-inverted"
];
const colorPaletteOptions = [
  ...themedColoringOptions,
  ...automaticColoringOptions,
  ...sequentialColoringOptions,
  ...divergingColoringOptions
];
const colorPaletteModeOptions = ["single-color", "multi-color"];
function isCustomColorPalette(colorPalette) {
  return typeof colorPalette === "object" && colorPalette !== null && (!Array.isArray(colorPalette) || colorPalette.length > 0);
}
function isCustomColorRecordPalette(colorPalette) {
  return !Array.isArray(colorPalette);
}
const shouldApplyAutomatically = (colorPalette) => {
  return automaticColoringOptions.includes(colorPalette);
};
const logStatusMap = {
  error: Colors.Charts.Logstatus.Error.Default,
  warning: Colors.Charts.Logstatus.Warning.Default,
  info: Colors.Charts.Logstatus.Info.Default,
  none: Colors.Charts.Logstatus.None.Default,
  warn: Colors.Charts.Logstatus.Warning.Default
  // alias for warning
};
const logLevelMap = {
  emergency: Colors.Charts.Loglevel.Emergency.Default,
  error: Colors.Charts.Loglevel.Error.Default,
  alert: Colors.Charts.Loglevel.Alert.Default,
  critical: Colors.Charts.Loglevel.Critical.Default,
  severe: Colors.Charts.Loglevel.Severe.Default,
  warning: Colors.Charts.Loglevel.Warning.Default,
  notice: Colors.Charts.Loglevel.Notice.Default,
  info: Colors.Charts.Loglevel.Info.Default,
  debug: Colors.Charts.Loglevel.Debug.Default,
  none: Colors.Charts.Loglevel.None.Default,
  warn: Colors.Charts.Loglevel.Warning.Default
  // alias for warning
};
const logStatusWithAlias = {
  error: Colors.Charts.Logstatus.Error,
  warning: Colors.Charts.Logstatus.Warning,
  info: Colors.Charts.Logstatus.Info,
  none: Colors.Charts.Logstatus.None,
  warn: Colors.Charts.Logstatus.Warning
  // alias for warning
};
const logLevelWithAlias = {
  emergency: Colors.Charts.Loglevel.Emergency,
  error: Colors.Charts.Loglevel.Error,
  alert: Colors.Charts.Loglevel.Alert,
  critical: Colors.Charts.Loglevel.Critical,
  severe: Colors.Charts.Loglevel.Severe,
  warning: Colors.Charts.Loglevel.Warning,
  notice: Colors.Charts.Loglevel.Notice,
  info: Colors.Charts.Loglevel.Info,
  debug: Colors.Charts.Loglevel.Debug,
  none: Colors.Charts.Loglevel.None,
  warn: Colors.Charts.Loglevel.Warning
  // alias for warning
};
const apdexMap = {
  excellent: Colors.Charts.Apdex.Excellent.Default,
  good: Colors.Charts.Apdex.Good.Default,
  fair: Colors.Charts.Apdex.Fair.Default,
  poor: Colors.Charts.Apdex.Poor.Default,
  unacceptable: Colors.Charts.Apdex.Unacceptable.Default
};
const securityRiskLevelMap = {
  muted: Colors.Charts.SecurityRiskLevel.Muted.Default,
  critical: Colors.Charts.SecurityRiskLevel.Critical.Default,
  high: Colors.Charts.SecurityRiskLevel.High.Default,
  medium: Colors.Charts.SecurityRiskLevel.Medium.Default,
  low: Colors.Charts.SecurityRiskLevel.Low.Default,
  none: Colors.Charts.SecurityRiskLevel.None.Default
};
const vulnerabilityStatusMap = {
  muted: Colors.Charts.VulnerabilityStatus.Muted.Default,
  resolved: Colors.Charts.VulnerabilityStatus.Resolved.Default,
  affected: Colors.Charts.VulnerabilityStatus.Affected.Default
};
const thresholdMap = {
  good: Colors.Charts.Threshold.Good.Default,
  warning: Colors.Charts.Threshold.Warning.Default,
  bad: Colors.Charts.Threshold.Bad.Default
};
const statusMap = {
  ideal: Colors.Charts.Status.Ideal.Default,
  good: Colors.Charts.Status.Good.Default,
  neutral: Colors.Charts.Status.Neutral.Default,
  warning: Colors.Charts.Status.Warning.Default,
  critical: Colors.Charts.Status.Critical.Default
};
export {
  apdexMap,
  automaticColoringOptions,
  colorPaletteModeOptions,
  colorPaletteOptions,
  divergingColoringOptions,
  isCustomColorPalette,
  isCustomColorRecordPalette,
  logLevelMap,
  logLevelWithAlias,
  logStatusMap,
  logStatusWithAlias,
  securityRiskLevelMap,
  sequentialColoringOptions,
  shouldApplyAutomatically,
  statusMap,
  themedColoringOptions,
  thresholdMap,
  vulnerabilityStatusMap
};
