// @dynatrace/strato-components/esm/charts/core/types/pointer-events.js
var PointerEventActions = /* @__PURE__ */ ((PointerEventActions2) => {
  PointerEventActions2["Idle"] = "idle";
  PointerEventActions2["Hover"] = "hover";
  PointerEventActions2["Click"] = "click";
  PointerEventActions2["DblClick"] = "dbl-click";
  PointerEventActions2["Tap"] = "tap";
  PointerEventActions2["DblTap"] = "dbl-tap";
  PointerEventActions2["DragStart"] = "drag:start";
  PointerEventActions2["DragMove"] = "drag:move";
  PointerEventActions2["DragEnd"] = "drag:end";
  PointerEventActions2["PinchStart"] = "pinch:start";
  PointerEventActions2["Pinch"] = "pinch";
  PointerEventActions2["PinchEnd"] = "pinch:end";
  PointerEventActions2["SwipeStart"] = "swipe:start";
  PointerEventActions2["SwipeMove"] = "swipe:move";
  PointerEventActions2["SwipeEnd"] = "swipe:end";
  PointerEventActions2["MiddleButtonDragStart"] = "middle-button-drag:start";
  PointerEventActions2["MiddleButtonDragMove"] = "middle-button-drag:move";
  PointerEventActions2["MiddleButtonDragEnd"] = "middle-button-drag:end";
  return PointerEventActions2;
})(PointerEventActions || {});
const isDragMove = (gesture) => gesture?.action === "drag:move" /* DragMove */;
const isMiddleButtonDragMove = (gesture) => gesture?.action === "middle-button-drag:move" /* MiddleButtonDragMove */;
const isPinchGesture = (gesture) => gesture?.action === "pinch" /* Pinch */;
const isPinchStartGesture = (gesture) => gesture?.action === "pinch:start" /* PinchStart */;
const isSwipeStart = (gesture) => gesture?.action === "swipe:start" /* SwipeStart */;
const isSwipeMove = (gesture) => gesture?.action === "swipe:move" /* SwipeMove */;
export {
  PointerEventActions,
  isDragMove,
  isMiddleButtonDragMove,
  isPinchGesture,
  isPinchStartGesture,
  isSwipeMove,
  isSwipeStart
};
