// @dynatrace/strato-components/esm/charts/core/types/type-guards.js
function isStatic(data) {
  return data !== null && !Array.isArray(data) && "value" in data;
}
function isStaticRange(data) {
  return data !== null && !Array.isArray(data) && "min" in data && "max" in data;
}
function isDynamic(data) {
  return data !== null && Array.isArray(data) && data[0] && "value" in data[0] && "date" in data[0];
}
function isDynamicNumeric(data) {
  return data !== null && Array.isArray(data) && data[0] && "value" in data[0] && "point" in data[0];
}
function isDynamicRange(data) {
  return data !== null && Array.isArray(data) && data[0] && "min" in data[0] && "max" in data[0] && "date" in data[0];
}
function isDynamicNumericRange(data) {
  return data !== null && Array.isArray(data) && data[0] && "min" in data[0] && "max" in data[0] && "point" in data[0];
}
function isPointThreshold(y0, y1) {
  return y0.valueOf() === y1.valueOf();
}
function isBoundedRange(data) {
  return data !== null && !Array.isArray(data) && "value" in data;
}
export {
  isBoundedRange,
  isDynamic,
  isDynamicNumeric,
  isDynamicNumericRange,
  isDynamicRange,
  isPointThreshold,
  isStatic,
  isStaticRange
};
