// @dynatrace/strato-components/charts/core/components/legend/categorical/context/LegendGridSettingsContext.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var LegendGridSettingsContext_exports = {};
__export(LegendGridSettingsContext_exports, {
  LegendGridSettingsContext: () => LegendGridSettingsContext,
  useLegendGridSettings: () => useLegendGridSettings
});
module.exports = __toCommonJS(LegendGridSettingsContext_exports);
var import_react = require("react");
const LegendGridSettingsContext = (0, import_react.createContext)({
  maxItemWidth: 0,
  canFitAllItemsWithoutScrollbar: false,
  hasVirtual: false,
  longestSeries: 0,
  maxColumns: 0,
  maxRows: 0,
  minItemWidth: 0
});
const useLegendGridSettings = () => (0, import_react.useContext)(LegendGridSettingsContext);
LegendGridSettingsContext["displayName"] = "LegendGridSettingsContext";
