// @dynatrace/strato-components/esm/charts/core/components/legend/range/utils/labels-collision.js
import { TEXT_ELLIPSIS } from "../constants.js";
import { getHorizontalLabelPosition } from "./get-horizontal-label-position.js";
import { TRUNCATE_LABEL } from "../../../../utils/text-truncation.js";
import { charsFontWidth12pixelsSize500Map } from "../../../text-ellipsis/chars-font-width-map.js";
import { calculateWidth } from "../../../text-ellipsis/truncate-string.js";
const calculateOptimalPlacement = (labelStart, labelEnd, startBound, endBound) => {
  if (labelStart < startBound) {
    const gap = startBound - labelStart;
    return labelStart + gap;
  }
  if (labelEnd > endBound) {
    const gap = endBound - labelEnd;
    return labelStart + gap;
  }
  return labelStart;
};
const calculateOptimalEllipsis = (text, startPosition, endPosition, truncationMode, startBound, endBound) => {
  let candidateStartPosX = startPosition;
  let candidateEndPosX = endPosition;
  let candidateLabel = text;
  let iteration = 0;
  do {
    candidateLabel = truncateLabel(candidateLabel, truncationMode);
    const newLabelWidth = calculateWidth(
      candidateLabel,
      charsFontWidth12pixelsSize500Map
    );
    const { x: newCandidateStartPosX } = getHorizontalLabelPosition(
      newLabelWidth,
      startBound,
      startBound + newLabelWidth,
      "start"
    );
    candidateStartPosX = newCandidateStartPosX;
    candidateEndPosX = candidateStartPosX + newLabelWidth;
    iteration++;
    if (iteration > text.length) {
      break;
    }
  } while (candidateStartPosX < startBound || candidateEndPosX > endBound);
  return { text: candidateLabel, x: candidateStartPosX };
};
const truncateLabel = (text, truncationMode) => {
  const filteredText = (text ?? "").replace(TEXT_ELLIPSIS, "");
  if (filteredText.length < 2) {
    return text;
  }
  return TRUNCATE_LABEL[truncationMode](filteredText);
};
const getLabelsPosition = (startLabel, endLabel, bounds) => {
  const startLabelWidth = calculateWidth(
    startLabel,
    charsFontWidth12pixelsSize500Map
  );
  const endLabelWidth = calculateWidth(
    endLabel,
    charsFontWidth12pixelsSize500Map
  );
  return {
    startLabel: { start: bounds[0], end: bounds[0] + startLabelWidth },
    endLabel: { start: bounds[1] - endLabelWidth, end: bounds[1] }
  };
};
export {
  calculateOptimalEllipsis,
  calculateOptimalPlacement,
  getLabelsPosition,
  truncateLabel
};
