// @dynatrace/strato-components/esm/filters/segment-selector/parts/selectionBar/SegmentSelectorTrigger.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import { ChevronDownSmallIcon, ContainerIcon } from "@dynatrace/strato-icons";
import * as styles from "./SegmentSelectorTrigger.sty.js";
import { Button } from "../../../../buttons/button/Button.js";
import { mergeProps as _mergeProps } from "../../../../core/utils/merge-props.js";
import { TextEllipsis } from "../../../../typography/text-ellipsis/TextEllipsis.js";
import { useSegmentsTriggerContext } from "../../hooks/useTriggerContext.js";
const SegmentSelectorTrigger = (props) => {
  const {
    textEllipsisRef,
    className: consumerClassName,
    ...remainingProps
  } = props;
  const { triggerProps, displayValue, isLoading, buttonRef, showError } = useSegmentsTriggerContext();
  return /* @__PURE__ */ jsxs(
    Button,
    {
      ref: buttonRef,
      loading: isLoading,
      variant: "emphasized",
      textAlign: "start",
      color: showError ? "critical" : "neutral",
      className: clsx(styles.trigger, consumerClassName),
      ..._mergeProps(triggerProps, remainingProps),
      children: [
        /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(ContainerIcon, {}) }),
        displayValue && /* @__PURE__ */ jsx(TextEllipsis, { ref: textEllipsisRef, children: displayValue }),
        /* @__PURE__ */ jsx(Button.Suffix, { children: /* @__PURE__ */ jsx(ChevronDownSmallIcon, {}) })
      ]
    }
  );
};
SegmentSelectorTrigger.displayName = "SegmentSelectorTrigger";
export {
  SegmentSelectorTrigger
};
