// @dynatrace/strato-components/esm/filters/segment-selector/parts/selectionBar/SegmentSelect.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import { forwardRef, useContext, useEffect, useRef } from "react";
import { useIntl } from "react-intl";
import * as styles from "./SegmentSelect.sty.js";
import { useMergeRefs } from "../../../../core/hooks/useMergeRefs.js";
import { isDev } from "../../../../core/utils/is-node-environment.js";
import { Select } from "../../../../forms/select/Select.js";
import { Text } from "../../../../typography/text/Text.js";
import {
  segmentsClient,
  useSegmentsCache
} from "../../contexts/SegmentsCache/SegmentsCache.js";
import {
  UISegmentsContext,
  useSegmentsUIContext
} from "../../contexts/SegmentsUI/SegmentsUIContext.js";
import { SEGMENTSELECTOR_MESSAGES } from "../../segmentMessages.js";
function getSafeSegmentName(segment, fallback) {
  return segment?.name && segment.name !== "" ? segment.name : fallback;
}
const SegmentSelect = forwardRef((props, forwardedRef) => {
  const {
    availableSegments,
    readOnly,
    triggerClassName,
    overlayWidth,
    segment: cachedSegment,
    autoFocus,
    isPlaceholder,
    id
  } = props;
  const intl = useIntl();
  const segmentSelectRef = useRef(null);
  const mergedRef = useMergeRefs([forwardedRef, segmentSelectRef]);
  const variableSelectRef = useRef(null);
  const { updateUISegment, removeUISegment } = useSegmentsUIContext();
  const { getSegment } = useSegmentsCache();
  const { uiSegments } = useContext(UISegmentsContext);
  const refIsDefined = segmentSelectRef.current !== null;
  useEffect(() => {
    segmentsClient.hasAccessRights && segmentSelectRef.current?.validate();
    if (uiSegments.length === 1 || autoFocus) {
      segmentSelectRef.current?.focus?.();
      return;
    }
    if (isPlaceholder && !autoFocus) {
      segmentSelectRef.current?.open();
    }
  }, [refIsDefined]);
  return /* @__PURE__ */ jsxs(
    Select,
    {
      ref: mergedRef,
      className: styles.segmentsSelectCSS,
      value: id,
      id,
      required: !isPlaceholder,
      clearable: false,
      readOnly: readOnly || cachedSegment?.temporarilyApplied,
      onOpenChange: (isOpen) => {
        if (isOpen) {
          return;
        }
        variableSelectRef.current?.open();
      },
      onChange: (segmentUid) => {
        if (segmentUid) {
          getSegment(segmentUid).then((newSegment) => {
            updateUISegment(
              { id },
              {
                ...newSegment.segment,
                variables: !newSegment.segment.variables ? void 0 : []
              }
            );
          }).catch((error) => {
            if (isDev()) {
              console.log(
                `SegmentSelector: fetching the segment with id ${id} failed.`,
                error
              );
            }
            removeUISegment(id);
          });
        }
      },
      "data-dt-name": "segment-select",
      children: [
        /* @__PURE__ */ jsx(
          Select.Trigger,
          {
            width: "full",
            className: clsx(triggerClassName, {
              [styles.segmentsSelectTriggerCSS]: cachedSegment?.segment?.variables
            }),
            placeholder: intl.formatMessage(
              SEGMENTSELECTOR_MESSAGES.segmentsTriggerPlaceholder
            )
          }
        ),
        /* @__PURE__ */ jsx(Select.Filter, {}),
        /* @__PURE__ */ jsxs(Select.Content, { showSelectedOptionsFirst: true, width: overlayWidth, children: [
          /* @__PURE__ */ jsx(Select.EmptyState, { children: availableSegments.length === 0 ? intl.formatMessage(
            SEGMENTSELECTOR_MESSAGES.segmentsSelectEmptyState
          ) : intl.formatMessage(
            SEGMENTSELECTOR_MESSAGES.segmentsSelectEmptyStateSearch
          ) }),
          (id && !isPlaceholder || cachedSegment?.segment) && /* @__PURE__ */ jsx(
            Select.Option,
            {
              value: id,
              textValue: getSafeSegmentName(cachedSegment?.segment, id),
              children: /* @__PURE__ */ jsx(Text, { children: getSafeSegmentName(cachedSegment?.segment, id) })
            },
            id
          ),
          availableSegments.map((segmentOption) => /* @__PURE__ */ jsx(
            Select.Option,
            {
              value: segmentOption.id,
              textValue: getSafeSegmentName(segmentOption, segmentOption.id),
              children: /* @__PURE__ */ jsx(Text, { children: getSafeSegmentName(segmentOption, segmentOption.id) })
            },
            segmentOption.id
          ))
        ] })
      ]
    }
  );
});
export {
  SegmentSelect
};
