// @dynatrace/strato-components/esm/filters/segment-selector/parts/selectionBar/SelectionBar.js
import { jsx, jsxs } from "react/jsx-runtime";
import { useQuery } from "@tanstack/react-query";
import clsx from "clsx";
import { useContext, useRef } from "react";
import { FormattedMessage, useIntl } from "react-intl";
import { XmarkIcon } from "@dynatrace/strato-icons";
import {
  clearButtonCSS,
  noBorderRadiusRight,
  segmentsSelectGridCSS
} from "./SegmentSelect.sty.js";
import { SegmentSelect } from "./SegmentSelect.js";
import { ValueSelect } from "./ValueSelect.js";
import { Button } from "../../../../buttons/button/Button.js";
import { useOverlayStacking } from "../../../../core/hooks/useOverlayStacking.js";
import { FormFieldMessages } from "../../../../forms/forms-core/components/form-field-messages/FormFieldMessages.js";
import { useFormFieldMessagesContext } from "../../../../forms/forms-core/hooks/useFormFieldMessagesContext.js";
import { Grid } from "../../../../layouts/grid/Grid.js";
import { Tooltip } from "../../../../overlays/tooltip/Tooltip.js";
import {
  segmentsClient,
  useSegmentsCache
} from "../../contexts/SegmentsCache/SegmentsCache.js";
import {
  UISegmentsContext,
  useSegmentsUIContext
} from "../../contexts/SegmentsUI/SegmentsUIContext.js";
import { getErrorKey } from "../../contexts/SegmentsUI/segmentValidationHelper.js";
import { SEGMENTSELECTOR_MESSAGES } from "../../segmentMessages.js";
import * as listStyles from "../../SegmentSelector.sty.js";
function getVariables(segment) {
  if (segment?.variables && segment.variables.length > 0) {
    return segment.variables[0].values;
  }
  return void 0;
}
function getLayout(hasVariables, hasClearButton) {
  if (!hasVariables && !hasClearButton) {
    return "no-variables-readonly";
  }
  if (!hasVariables && hasClearButton) {
    return "no-variables";
  }
  if (hasVariables && !hasClearButton) {
    return "full-readonly";
  }
  return "full";
}
function getErrorMessageFromKey(errorKey) {
  switch (errorKey) {
    case "unavailable":
      return SEGMENTSELECTOR_MESSAGES.unavailableSegmentError;
    case "needs-update":
      return SEGMENTSELECTOR_MESSAGES.removedVariablesError;
    case "needs-variables":
      return SEGMENTSELECTOR_MESSAGES.valueMissingError;
    default:
      return null;
  }
}
const ErrorMessages = ({ uiSegment }) => {
  const {
    data: errorKey,
    isLoading,
    error
  } = useQuery(
    {
      queryKey: ["errorKey", uiSegment],
      queryFn: () => getErrorKey(uiSegment)
    },
    segmentsClient.queryClient
  );
  const { arrayIdentifierErrorIds } = useContext(UISegmentsContext);
  const hasArrayIdentifierError = arrayIdentifierErrorIds?.includes(
    uiSegment.id
  );
  if (segmentsClient.hasAccessRights === false || isLoading || error) {
    return null;
  }
  const errorMessage = getErrorMessageFromKey(errorKey);
  return /* @__PURE__ */ jsx(FormFieldMessages, { children: (messages) => {
    if (hasArrayIdentifierError) {
      return /* @__PURE__ */ jsx(
        FormFieldMessages.Item,
        {
          validityStateKey: "customError",
          variant: "error",
          children: /* @__PURE__ */ jsx(
            FormattedMessage,
            {
              ...SEGMENTSELECTOR_MESSAGES.arrayIdentifierError
            }
          )
        },
        "array-identifier"
      );
    }
    if (errorKey) {
      return /* @__PURE__ */ jsx(
        FormFieldMessages.Item,
        {
          validityStateKey: "customError",
          variant: "error",
          children: /* @__PURE__ */ jsx(FormattedMessage, { ...errorMessage })
        },
        errorKey
      );
    }
    return messages.map((props) => {
      if (props.controlId !== uiSegment.id && props.validityStateKey === "valueMissing") {
        return /* @__PURE__ */ jsx(
          FormFieldMessages.Item,
          {
            ...props,
            children: /* @__PURE__ */ jsx(
              FormattedMessage,
              {
                ...SEGMENTSELECTOR_MESSAGES.valueMissingError
              }
            )
          },
          `${props.controlId}-valueMissing`
        );
      }
      return /* @__PURE__ */ jsx(
        FormFieldMessages.Item,
        {
          ...props
        },
        `${props.controlId}-${props.validityStateKey}`
      );
    });
  } });
};
const SelectionBar = (props) => {
  const {
    availableSegments,
    clearable,
    readOnly,
    autoFocus,
    uiSegment,
    returnFocusTo,
    segment: propSegment
  } = props;
  const intl = useIntl();
  const gridRef = useRef(null);
  const isPlaceholder = uiSegment.isPlaceholder ?? false;
  const id = uiSegment.id;
  const variables = getVariables(uiSegment);
  const { removeUISegment } = useSegmentsUIContext();
  const { state } = useFormFieldMessagesContext();
  const { getSegment } = useSegmentsCache();
  const { data: fetchedSegment } = useQuery(
    {
      queryKey: ["getSegment", id],
      queryFn: () => getSegment(id),
      enabled: segmentsClient.hasAccessRights !== false && !isPlaceholder && propSegment === void 0,
      retryOnMount: false
    },
    segmentsClient.queryClient
  );
  const segment = propSegment ?? fetchedSegment;
  const showDeleteButton = clearable && !readOnly && segmentsClient.hasAccessRights !== false;
  const stackingContext = useOverlayStacking(void 0, "overlay");
  return /* @__PURE__ */ jsxs("li", { className: listStyles.segmentSelectorListItem, children: [
    /* @__PURE__ */ jsx(
      Tooltip,
      {
        text: intl.formatMessage(
          SEGMENTSELECTOR_MESSAGES.unlistedSegmentTooltip
        ),
        style: { zIndex: Number(stackingContext.style.zIndex) + 1 },
        disabled: !segment?.temporarilyApplied,
        children: /* @__PURE__ */ jsxs(
          Grid,
          {
            ref: gridRef,
            className: segmentsSelectGridCSS({
              layout: getLayout(
                !!segment?.segment?.variables,
                !!showDeleteButton
              )
            }),
            children: [
              /* @__PURE__ */ jsx(
                SegmentSelect,
                {
                  readOnly,
                  availableSegments,
                  segment,
                  id,
                  overlayWidth: `${gridRef.current?.clientWidth}px`,
                  triggerClassName: clsx({
                    [noBorderRadiusRight]: showDeleteButton || segment?.segment?.variables
                  }),
                  isPlaceholder,
                  autoFocus
                }
              ),
              segment?.segment?.variables && /* @__PURE__ */ jsx(
                ValueSelect,
                {
                  readOnly,
                  segment,
                  variables,
                  triggerClassName: clsx({
                    [noBorderRadiusRight]: clearable && !readOnly
                  })
                }
              ),
              showDeleteButton && /* @__PURE__ */ jsx(
                Button,
                {
                  onClick: () => {
                    removeUISegment(id);
                    returnFocusTo.current?.focus();
                  },
                  color: state.showError ? "critical" : "neutral",
                  variant: "emphasized",
                  className: clearButtonCSS,
                  "aria-label": intl.formatMessage(
                    SEGMENTSELECTOR_MESSAGES.removeSegmentButton
                  ),
                  "data-dt-name": "Remove",
                  children: /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(XmarkIcon, {}) })
                }
              )
            ]
          }
        )
      }
    ),
    /* @__PURE__ */ jsx(ErrorMessages, { uiSegment })
  ] });
};
export {
  ErrorMessages,
  SelectionBar
};
