// @dynatrace/strato-components/esm/filters/segment-selector/parts/selectionBar/ValueSelect.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import { isArray } from "lodash-es";
import { useContext, useEffect, useMemo, useRef } from "react";
import { FormattedMessage, useIntl } from "react-intl";
import Colors from "@dynatrace/strato-design-tokens/colors";
import { useDql } from "@dynatrace-sdk/react-hooks";
import * as styles from "./SegmentSelect.sty.js";
import { StratoLoggerContext } from "../../../../core/sdk/StratoLoggerContext.js";
import { useFormFieldMessagesContext } from "../../../../forms/forms-core/hooks/useFormFieldMessagesContext.js";
import { FormFieldMessagesAction } from "../../../../forms/forms-core/state/formfield-messages-state-reducer.js";
import { Select } from "../../../../forms/select/Select.js";
import { Grid } from "../../../../layouts/grid/Grid.js";
import { Link } from "../../../../typography/link/Link.js";
import { Text } from "../../../../typography/text/Text.js";
import { TextEllipsis } from "../../../../typography/text-ellipsis/TextEllipsis.js";
import {
  UISegmentsContext,
  useSegmentsUIContext
} from "../../contexts/SegmentsUI/SegmentsUIContext.js";
import { SEGMENTSELECTOR_MESSAGES } from "../../segmentMessages.js";
import {
  convertQueryRecordToStoreSegmentVariables,
  getReadableValue
} from "../../utils.js";
import { getSegmentHref } from "../segmentManagementHref.js";
const dtClientContext = JSON.stringify({
  source: "SegmentSelector",
  package: "@dynatrace/strato-components"
});
const ValueSelect = (props) => {
  const { triggerClassName, readOnly, segment, variables = [] } = props;
  const variableSelectRef = useRef(null);
  const { focusedSegmentId, isTouched } = useContext(UISegmentsContext);
  const { updateUISegment, setArrayIdentifierError } = useSegmentsUIContext();
  const { dispatch: messagesDispatch } = useFormFieldMessagesContext();
  const intl = useIntl();
  const stratoLogger = useContext(StratoLoggerContext);
  const query = segment?.segment?.variables !== void 0 && segment?.segment?.variables?.type === "query" ? segment.segment.variables.value : "";
  const {
    data,
    isLoading: isDqlLoading,
    forceRefetch
  } = useDql({
    query,
    maxResultRecords: 1e4,
    maxResultBytes: 5242880,
    // 5 MB
    dtClientContext
  });
  useEffect(() => {
    if (!isDqlLoading && data?.records?.length === 0) {
      stratoLogger?.info(
        "Failed to retrieve segment variables: query returned no records.",
        null,
        {
          query: query.split("|")[0],
          // log query until first pipe to avoid reporting sensitive data.
          id: segment?.segment?.id
        }
      );
    }
  }, [query, isDqlLoading]);
  const unavailableVariables = useMemo(() => {
    if (data) {
      const firstKey = Object.keys(data.records?.[0] ?? {})[0];
      const dataRecordIds = data.records.map((record) => record?.[firstKey]);
      return variables?.filter((variable) => !dataRecordIds.includes(variable));
    }
  }, [data, variables]);
  const availableVariables = useMemo(() => {
    if (data?.records) {
      let hasArrayIdentifier = false;
      const recordsArray = data?.records.map((singleRow) => {
        const firstKey = singleRow ? Object.keys(singleRow)[0] : null;
        if (!singleRow || !firstKey) {
          return null;
        }
        if (isArray(singleRow[firstKey])) {
          hasArrayIdentifier = true;
        }
        return `${singleRow[firstKey]}`;
      }).filter((value) => value !== "null" && value !== null);
      queueMicrotask(() => {
        if (hasArrayIdentifier) {
          messagesDispatch?.({
            type: FormFieldMessagesAction.ADDMESSAGE,
            payload: {
              variant: "error",
              messageType: "formcontrol",
              id: "array-identifier",
              validityStateKey: "customError",
              message: /* @__PURE__ */ jsx(
                FormattedMessage,
                {
                  ...SEGMENTSELECTOR_MESSAGES.arrayIdentifierError
                }
              )
            }
          });
        }
        setArrayIdentifierError(hasArrayIdentifier, segment.segment.id);
      });
      return Array.from(new Set(recordsArray));
    }
  }, [data]);
  const shouldOpen = focusedSegmentId === segment.segment.id && variables.length === 0 && isTouched;
  useEffect(() => {
    shouldOpen && variableSelectRef.current?.open();
  }, [shouldOpen]);
  useEffect(() => {
    !isTouched && variableSelectRef.current?.validate();
  }, [isTouched, JSON.stringify(variables)]);
  return /* @__PURE__ */ jsxs(
    Select,
    {
      ref: variableSelectRef,
      multiple: true,
      className: styles.variablesSelectCSS,
      required: true,
      readOnly: readOnly || segment?.temporarilyApplied,
      clearable: false,
      onOpenChange: (isOpen) => {
        if (query.length > 0 && isOpen) {
          forceRefetch().catch(() => void 0);
        }
      },
      value: variables,
      onChange: (newVariables) => {
        if (!newVariables || newVariables.length === 0) {
          updateUISegment(
            {
              ...segment.segment,
              variables: []
            },
            {
              ...segment.segment,
              variables: []
            }
          );
          return;
        }
        const firstKey = Object.keys(data?.records[0] ?? {})[0];
        const dataEnrichedVariables = newVariables.map((variable) => {
          return data?.records.find((singleRow) => {
            if (isArray(singleRow?.[firstKey])) {
              const values = singleRow?.[firstKey].includes(variable);
              return values;
            }
            return singleRow && singleRow[firstKey] === variable;
          }) ?? null;
        }).filter((value) => value !== null);
        const segmentVariables = convertQueryRecordToStoreSegmentVariables(
          dataEnrichedVariables
        );
        if (segmentVariables) {
          updateUISegment(
            {
              ...segment.segment,
              variables: []
            },
            {
              ...segment.segment,
              variables: segmentVariables
            }
          );
        }
      },
      "data-dt-name": "segment-variable-select",
      children: [
        /* @__PURE__ */ jsx(
          Select.Trigger,
          {
            className: clsx(styles.variablesSelectTriggerCSS, triggerClassName),
            placeholder: intl.formatMessage(
              SEGMENTSELECTOR_MESSAGES.segmentsVariableTriggerPlaceholder
            ),
            children: variables?.length > 0 && /* @__PURE__ */ jsx(Select.DisplayValue, { children: /* @__PURE__ */ jsx(TextEllipsis, { children: variables.map((variable) => getReadableValue(variable, intl)).join(", ") }) })
          }
        ),
        /* @__PURE__ */ jsx(Select.Filter, {}),
        /* @__PURE__ */ jsxs(
          Select.Content,
          {
            loading: isDqlLoading,
            showSelectedOptionsFirst: true,
            width: "trigger",
            children: [
              !data && variables?.map((variable) => /* @__PURE__ */ jsx(
                Select.Option,
                {
                  value: variable,
                  textValue: getReadableValue(variable, intl) ?? void 0,
                  children: getReadableValue(variable, intl)
                },
                variable
              )),
              unavailableVariables?.map((unlistedVariable) => /* @__PURE__ */ jsx(
                Select.Option,
                {
                  value: unlistedVariable,
                  textValue: getReadableValue(unlistedVariable, intl) ?? void 0,
                  children: /* @__PURE__ */ jsxs(
                    Grid,
                    {
                      gridTemplateColumns: "auto auto",
                      width: "100%",
                      alignItems: "baseline",
                      children: [
                        /* @__PURE__ */ jsx(Text, { children: getReadableValue(unlistedVariable, intl) }),
                        /* @__PURE__ */ jsx(
                          Text,
                          {
                            textStyle: "small",
                            style: {
                              color: Colors.Text.Neutral.Subdued,
                              textAlign: "right",
                              textWrap: "nowrap"
                            },
                            children: intl.formatMessage(SEGMENTSELECTOR_MESSAGES.unavailable)
                          }
                        )
                      ]
                    }
                  )
                },
                unlistedVariable
              )),
              availableVariables?.map((variable) => /* @__PURE__ */ jsxs(
                Select.Option,
                {
                  value: variable,
                  textValue: getReadableValue(variable, intl) ?? void 0,
                  children: [
                    " ",
                    /* @__PURE__ */ jsx(Text, { children: getReadableValue(variable, intl) })
                  ]
                },
                variable
              )),
              /* @__PURE__ */ jsx(Select.EmptyState, { children: data?.records.length ? /* @__PURE__ */ jsx(
                FormattedMessage,
                {
                  id: "nIw/xeiGNJsj+a+5",
                  defaultMessage: "No matching values for this segment found.",
                  description: "Hint shown when searching for a segment variable that isn't in the list."
                }
              ) : /* @__PURE__ */ jsx(
                FormattedMessage,
                {
                  defaultMessage: "No values for this segment available. You can manage this segment in {link}.",
                  id: "72DlIwYKB9UTEzlA",
                  description: "Empty state message if no variables for the segment are available.",
                  values: {
                    link: /* @__PURE__ */ jsx(
                      Link,
                      {
                        href: getSegmentHref(segment.segment.id),
                        target: "_blank",
                        rel: "noopener noreferrer",
                        children: /* @__PURE__ */ jsx(
                          FormattedMessage,
                          {
                            defaultMessage: "settings",
                            id: "u55UYpLvXi//xGEn",
                            description: "Link text to segments app in empty state message if no variables for the segment are available."
                          }
                        )
                      }
                    )
                  }
                }
              ) })
            ]
          }
        )
      ]
    }
  );
};
export {
  ValueSelect
};
