// @dynatrace/strato-components/charts/pie/providers/ChartProviders.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ChartProviders_exports = {};
__export(ChartProviders_exports, {
  ChartProviders: () => ChartProviders
});
module.exports = __toCommonJS(ChartProviders_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_StoreProvider = require("./StoreProvider.js");
var import_ChartSeriesActionContext = require("../../core/components/series-action/ChartSeriesActionContext.js");
var import_ChartLegendContext = require("../context/ChartLegendContext.js");
var import_FormatterContext = require("../context/FormatterContext.js");
var import_LabelsContext = require("../context/LabelsContext.js");
var import_PieChartConfigContext = require("../context/PieChartConfigContext.js");
var import_PieChartRawDataContext = require("../context/PieChartRawDataContext.js");
var import_PieChartTooltipContext = require("../context/PieChartTooltipContext.js");
var import_ToolbarContext = require("../context/ToolbarContext.js");
const ChartProviders = (props) => {
  const { chartConfig, chartData, children } = props;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ChartSeriesActionContext.ChartSeriesActionContext.Provider,
    {
      value: {
        // We need a cast here as the provider expects unknown and the
        // react context will not be able to carry the generic anyways.
        chartSeriesActions: chartConfig.seriesActions
      },
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_PieChartRawDataContext.PieChartRawDataContext.Provider, { value: chartData, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_PieChartConfigContext.PieChartConfigContext.Provider, { value: chartConfig, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_PieChartTooltipContext.PieChartTooltipContext.Provider,
        {
          value: chartConfig.tooltip ?? import_PieChartTooltipContext.DEFAULT_TOOLTIP_CONFIG,
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_LabelsContext.LabelsContext.Provider,
            {
              value: {
                hidden: chartConfig.labels?.hidden,
                valueType: chartConfig.labels?.valueType
              },
              children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_FormatterContext.FormatterContext.Provider,
                {
                  value: {
                    formatter: chartConfig.formatter,
                    truncationMode: chartConfig.truncationMode
                  },
                  children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ChartLegendContext.ChartLegendContext.Provider, { value: chartConfig.legend, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ToolbarContext.ToolbarContext.Provider, { value: chartConfig.toolbar, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_StoreProvider.StoreProvider, { children }) }) })
                }
              )
            }
          )
        }
      ) }) })
    }
  );
};
