// @dynatrace/strato-components/esm/forms/date-time-range-picker/DateTimeRangePicker.js
import { jsx } from "react/jsx-runtime";
import {
  forwardRef,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState
} from "react";
import { DateTimeRangePickerRoot } from "./DateTimeRangePickerRoot.js";
import { useCallbackRef } from "../../core/hooks/useCallbackRef.js";
import { useMergeRefs } from "../../core/hooks/useMergeRefs.js";
import { getDateTimeRangePickerActions } from "../../core/time/timeframeStore/actions/dateTimeRangePicker.js";
import { createTimeframeStore } from "../../core/time/timeframeStore/createTimeframeStore.js";
import { DateTimeRangePickerStoreProvider } from "../../core/time/timeframeStore/timeframeStoreProvider.js";
import { useDateTimeMessages } from "../../core/time/useDateTimeMessages.js";
import { useMinMax, useMinMaxStoreSync } from "../../core/time/useMinMax.js";
import { validate } from "../../core/time/validation.js";
import { getTimeOffsetFromInitialValue } from "../calendar/utils/helpers.js";
import { useFormFieldMessagesContext } from "../forms-core/hooks/useFormFieldMessagesContext.js";
import { useSyncedFormControlProps } from "../forms-core/hooks/useSyncedFormControlProps.js";
import {
  FormFieldMessagesAction
} from "../forms-core/state/formfield-messages-state-reducer.js";
import {
  GRAIL_MAX_DATE,
  GRAIL_MIN_DATE
} from "../forms-core/time/date-time.js";
import {
  removePrecisionFromTimeFrame,
  mapToPartialTimeframe
} from "../forms-core/time/timeParser.js";
function useSyncPropsAndState(store, memoedValue, isControlled, precision, formatString, required, formFieldMessagesDispatch, dateTimeRangePickerRef) {
  useEffect(() => {
    isControlled && store.setState((state) => {
      const error = validate(memoedValue ?? null, state.min, state.max, {
        formMessageDispatch: formFieldMessagesDispatch,
        precision: state.precision,
        showError: false,
        checkOrder: true,
        formerError: state.calendarError,
        ref: dateTimeRangePickerRef,
        expectedFormat: state.formatString,
        valueMode: "dateTimePicker",
        required: state.required
      });
      const value = removePrecisionFromTimeFrame(
        memoedValue,
        state.precision
      );
      const timeOffset = getTimeOffsetFromInitialValue(
        value,
        "range",
        state.min,
        state.max,
        precision
      );
      return {
        ...state,
        dateTimeRangePickerTimeframe: value,
        calendarValue: value,
        calendarError: error,
        dateTimeRangePickerError: error,
        timeOffset
      };
    });
  }, [
    formFieldMessagesDispatch,
    isControlled,
    memoedValue,
    store,
    precision,
    dateTimeRangePickerRef
  ]);
  useEffect(() => {
    store.setState((state) => {
      const value = removePrecisionFromTimeFrame(
        state.dateTimeRangePickerTimeframe,
        precision
      );
      return {
        ...state,
        precision,
        formatString,
        dateTimeRangePickerTimeframe: value,
        calendarValue: value,
        required
      };
    });
  }, [precision, formatString, store, required]);
}
const DateTimeRangePicker = forwardRef(
  ({
    value: propValue,
    defaultValue: propDefaultValue,
    onChange,
    min: propMin = GRAIL_MIN_DATE,
    max: propMax = GRAIL_MAX_DATE,
    precision = "minutes",
    disabled: propDisabled,
    required: propRequired,
    readOnly: propReadOnly,
    id: propId,
    "aria-disabled": propAriaDisabled,
    onBlur,
    ...remainingProps
  }, forwardedRef) => {
    const [isControlled] = useState(propValue !== void 0);
    const onChangeProp = useCallbackRef(onChange);
    const memoedValue = useMemo(
      () => mapToPartialTimeframe(propValue, precision),
      [precision, propValue]
    );
    const memoedDefaultValue = useMemo(
      () => mapToPartialTimeframe(propDefaultValue, precision),
      /** We're ignoring updates to defaultValue as it only considers the initial state. */
      // eslint-disable-next-line react-hooks/exhaustive-deps
      []
    );
    const { min, max } = useMinMax(propMin, propMax);
    const calendarRef = useRef(null);
    const dateTimeFromRef = useRef(null);
    const dateTimeRangePickerRef = useRef(null);
    const { dispatch: formFieldMessagesDispatch } = useFormFieldMessagesContext();
    const { formatString } = useDateTimeMessages({ precision });
    const { disabled, required, id, ariaDisabled, readOnly } = useSyncedFormControlProps({
      propDisabled,
      propRequired,
      propId,
      propAriaDisabled,
      propReadOnly
    });
    const [initialState] = useState(() => {
      const value = isControlled ? memoedValue : memoedDefaultValue;
      const timeOffset = getTimeOffsetFromInitialValue(
        value,
        "range",
        min,
        max,
        precision
      );
      const error = validate(value, min, max, {
        formMessageDispatch: formFieldMessagesDispatch,
        precision,
        showError: false,
        checkOrder: true,
        ref: dateTimeRangePickerRef,
        expectedFormat: formatString,
        valueMode: "dateTimePicker",
        required
      });
      return {
        calendarValue: value,
        dateTimeRangePickerTimeframe: value,
        min,
        max,
        precision,
        calendarRef,
        dateTimeFromRef,
        dateTimeRangePickerRef,
        calendarError: error,
        dateTimeRangePickerError: error,
        timeOffset,
        formatString,
        required
      };
    });
    const [store] = useState(
      () => createTimeframeStore(
        initialState,
        getDateTimeRangePickerActions(formFieldMessagesDispatch, onChangeProp)
      )
    );
    useSyncPropsAndState(
      store,
      memoedValue,
      isControlled,
      precision,
      formatString,
      required,
      formFieldMessagesDispatch,
      dateTimeRangePickerRef
    );
    useMinMaxStoreSync(
      store,
      min,
      max,
      useCallback(
        (state, min2, max2) => {
          const error = validate(state.dateTimeRangePickerTimeframe, min2, max2, {
            formMessageDispatch: formFieldMessagesDispatch,
            formerError: state.calendarError,
            precision: state.precision,
            showError: false,
            checkOrder: true,
            ref: dateTimeRangePickerRef,
            expectedFormat: state.formatString,
            valueMode: "dateTimePicker",
            required: state.required
          });
          return { calendarError: error, dateTimeRangePickerError: error };
        },
        [formFieldMessagesDispatch]
      )
    );
    const mergedRef = useMergeRefs([
      dateTimeRangePickerRef,
      forwardedRef
    ]);
    return /* @__PURE__ */ jsx(DateTimeRangePickerStoreProvider, { store, children: /* @__PURE__ */ jsx(
      DateTimeRangePickerRoot,
      {
        ...remainingProps,
        id,
        ref: mergedRef,
        "aria-disabled": ariaDisabled,
        disabled,
        required,
        readOnly,
        onBlur: (event) => {
          onBlur?.(event);
          if (!event.target.contains(document.activeElement)) {
            formFieldMessagesDispatch?.({
              type: FormFieldMessagesAction.SETISTOUCHED,
              payload: { isTouched: true }
            });
          }
        }
      }
    ) });
  }
);
export {
  DateTimeRangePicker
};
