// @dynatrace/strato-components/esm/forms/date-time-range-picker/DateTimeRangePickerRoot.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import {
  forwardRef,
  useCallback,
  useImperativeHandle,
  useMemo,
  useRef,
  useState
} from "react";
import { useIntl } from "react-intl";
import { XmarkIcon } from "@dynatrace/strato-icons";
import { DateRangeSelector } from "./DateRangeSelector.js";
import { DateTimeInput } from "./DateTimeInput.js";
import * as styles from "./DateTimeRangePicker.sty.js";
import { isExpressionTriggerKey } from "./expression-trigger-helpers.js";
import { useDateTimeRangePicker } from "./hooks/useDateTimeRangePicker.js";
import { onTimeframeCopyCapture } from "./selection-copy-helpers.js";
import {
  peekPreRestoreEndpoints
} from "./selection-helpers.js";
import { useSelectionInteractionHandlers } from "./useSelectionInteractionHandlers.js";
import { Button } from "../../buttons/button/Button.js";
import { useAriaLabelingProps } from "../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useFocusRing } from "../../core/styles/useFocusRing.js";
import { useDateTimeRangePickerStore } from "../../core/time/timeframeStore/useTimeframeStore.js";
import { validate } from "../../core/time/validation.js";
import { Flex } from "../../layouts/flex/Flex.js";
import { inputGroupClassName } from "../../layouts/input-group/InputGroup.sty.js";
import { InputGroup } from "../../layouts/input-group/InputGroup.js";
import { fieldTokens } from "../../styles/field.sty.js";
import { textStyle } from "../../styles/textStyle.sty.js";
import {
  clearIconButton,
  dateTimePicker
} from "../date-time-picker/DateTimePicker.sty.js";
import { focusDateTimeInput } from "../date-time-picker/utils/focusDateTimeInput.js";
import { FormFieldMessageContext } from "../forms-core/contexts/FormFieldMessageContext.js";
import { useFormFieldMessagesContext } from "../forms-core/hooks/useFormFieldMessagesContext.js";
import { useSyncedFormControlProps } from "../forms-core/hooks/useSyncedFormControlProps.js";
import { FormFieldMessagesAction } from "../forms-core/state/formfield-messages-state-reducer.js";
function isInputInSelectionOrFocused(ref, preRestoreEndpoints) {
  const element = ref.current?.element;
  if (!element) {
    return false;
  }
  if (element.contains(document.activeElement)) {
    return true;
  }
  const selection = window.getSelection();
  if (selection && selection.rangeCount > 0 && !selection.isCollapsed) {
    if (selection.getRangeAt(0).intersectsNode(element)) {
      return true;
    }
  }
  return !!preRestoreEndpoints && (element.contains(preRestoreEndpoints.anchorNode) || element.contains(preRestoreEndpoints.focusNode));
}
function focusExpressionInputAtEnd(ref) {
  const expressionElement = ref.current?.expressionInputRef;
  if (!expressionElement) {
    return;
  }
  expressionElement.focus();
  const selection = window.getSelection();
  if (!selection) {
    return;
  }
  const range = document.createRange();
  range.selectNodeContents(expressionElement);
  range.collapse(false);
  selection.removeAllRanges();
  selection.addRange(range);
}
function getOffsetWithinElement(element, container, offset) {
  const range = document.createRange();
  range.selectNodeContents(element);
  range.setEnd(container, offset);
  return range.toString().length;
}
const DateTimeRangePickerRoot = forwardRef((props, forwardedRef) => {
  const {
    id: propId,
    disabled: propDisabled,
    readOnly = false,
    name,
    style: consumerStyle,
    className: consumerClassName,
    required: propRequired,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "aria-disabled": ariaDisabledProp,
    placeholder: placeholderProp,
    onBlur,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps] = useBehavioralTrackingProps(remainingProps);
  const ariaProps = useAriaLabelingProps(props);
  const COMPONENT_NAME = "DateTimeRangePicker";
  const rootRef = useRef(null);
  const touchedContainerRef = useRef(null);
  const dateTimeToRef = useRef(null);
  const dateRangeSelectorRef = useRef(null);
  const [hasInputContent, setHasInputContent] = useState(false);
  const setValue = useDateTimeRangePickerStore(
    (state) => state.setDateTimeRangePickerTimeframe
  );
  const value = useDateTimeRangePickerStore(
    (state) => state.dateTimeRangePickerTimeframe
  );
  const dateTimeFromRef = useDateTimeRangePickerStore(
    (state) => state.dateTimeFromRef
  );
  const precision = useDateTimeRangePickerStore((state) => state.precision);
  const min = useDateTimeRangePickerStore((state) => state.min);
  const max = useDateTimeRangePickerStore((state) => state.max);
  const formatString = useDateTimeRangePickerStore(
    (state) => state.formatString
  );
  const { disabled, required, id, ariaDisabled } = useSyncedFormControlProps({
    propDisabled,
    propRequired,
    propId,
    propAriaDisabled: ariaDisabledProp
  });
  const disabledOrAriaDisabled = disabled || ariaDisabled || false;
  const { placeholder, onPasteCapture, onUndoPasteKeyDown } = useDateTimeRangePicker({
    placeholder: placeholderProp,
    readOnly,
    dateTimeFromRef,
    dateTimeToRef
  });
  const intl = useIntl();
  const {
    state: { showError },
    dispatch: formMessageDispatch
  } = useFormFieldMessagesContext();
  const lastValidationErrorRef = useRef();
  useImperativeHandle(
    forwardedRef,
    () => ({
      element: rootRef.current,
      inputRef: {
        from: dateTimeFromRef.current?.inputRef,
        to: dateTimeToRef.current?.inputRef
      },
      open: () => dateRangeSelectorRef.current?.open(),
      close: () => dateRangeSelectorRef.current?.close(),
      focus: () => dateTimeFromRef.current?.focus?.(),
      getCurrentValue: () => ({
        from: value?.from ?? null,
        to: value?.to ?? null
      }),
      validate() {
        lastValidationErrorRef.current = validate(value, min, max, {
          formMessageDispatch,
          formerError: lastValidationErrorRef.current,
          precision,
          showError: true,
          checkOrder: true,
          ref: rootRef,
          expectedFormat: formatString,
          valueMode: "dateTimePicker",
          required
        });
        formMessageDispatch?.({
          type: FormFieldMessagesAction.SETISTOUCHED,
          payload: { isTouched: true }
        });
      }
    }),
    [
      value,
      min,
      max,
      formMessageDispatch,
      precision,
      formatString,
      required,
      dateTimeFromRef
    ]
  );
  const { focusClassName } = useFocusRing({
    focusWithin: true,
    ignoreModality: true,
    isMinimal: false,
    disabled,
    color: showError ? "critical" : "neutral"
  });
  const dateTimeFromLabel = intl.formatMessage({
    defaultMessage: "DateTime from",
    id: "fMy6q/KKOqbvCA3H",
    description: "Textbox to enter a relative or absolute timeframe from value."
  });
  const dateTimeToLabel = intl.formatMessage({
    defaultMessage: "DateTime to",
    id: "uuhhc3nEZptwtV2E",
    description: "Textbox to enter a relative or absolute timeframe to value."
  });
  const clearButtonLabel = intl.formatMessage({
    defaultMessage: "Clear date and time range",
    id: "zX4r/Y8gJL1C8+5e",
    description: "Aria label for the DateTimeRangePicker clear button."
  });
  const formFieldMessageContextValue = useMemo(
    () => ({
      state: {
        customMessages: [],
        formControlMessages: [],
        showError
      }
    }),
    [showError]
  );
  const { onMouseDown, onKeyDown: handleSelectionKeyDown } = useSelectionInteractionHandlers(
    touchedContainerRef,
    useCallback(() => {
      focusDateTimeInput(null, dateTimeFromRef, dateTimeToRef);
    }, [dateTimeFromRef]),
    dateTimeFromRef,
    dateTimeToRef
  );
  const onKeyDown = useCallback(
    (event) => {
      if (disabled || readOnly) {
        handleSelectionKeyDown(event);
        return;
      }
      const preRestoreEndpoints = peekPreRestoreEndpoints(
        event.currentTarget
      );
      const fromInputIsIso = dateTimeFromRef.current?.element?.getAttribute("data-valuetype") === "iso8601";
      const toInputIsIso = dateTimeToRef.current?.element?.getAttribute("data-valuetype") === "iso8601";
      const bothInSelection = isInputInSelectionOrFocused(dateTimeFromRef, preRestoreEndpoints) && isInputInSelectionOrFocused(dateTimeToRef, preRestoreEndpoints);
      if (bothInSelection && isExpressionTriggerKey(event)) {
        setValue(null);
        if (dateTimeToRef.current?.expressionInputRef) {
          dateTimeToRef.current.expressionInputRef.textContent = "";
        }
      }
      const switchFromToExpression = isInputInSelectionOrFocused(dateTimeFromRef, preRestoreEndpoints) && fromInputIsIso;
      const switchToToExpression = isInputInSelectionOrFocused(dateTimeToRef, preRestoreEndpoints) && toInputIsIso;
      if (isExpressionTriggerKey(event) && (switchFromToExpression || switchToToExpression)) {
        const char = event.key;
        const primaryRef = switchFromToExpression ? dateTimeFromRef : dateTimeToRef;
        primaryRef.current?.setExpressionValue(char);
        requestAnimationFrame(() => {
          focusExpressionInputAtEnd(primaryRef);
        });
        event.preventDefault();
        event.stopPropagation();
        return;
      }
      if (isInputInSelectionOrFocused(dateTimeFromRef, preRestoreEndpoints) && isInputInSelectionOrFocused(dateTimeToRef, preRestoreEndpoints) && !fromInputIsIso && !toInputIsIso && event.key.length === 1 && !event.ctrlKey && !event.metaKey && !event.altKey) {
        const char = event.key;
        const fromExpressionElement = dateTimeFromRef.current?.expressionInputRef;
        const toExpressionElement = dateTimeToRef.current?.expressionInputRef;
        const selection = window.getSelection();
        if (fromExpressionElement && toExpressionElement && selection && selection.rangeCount > 0 && !selection.isCollapsed) {
          const range = selection.getRangeAt(0);
          const fromValue = fromExpressionElement.textContent ?? "";
          const toValue = toExpressionElement.textContent ?? "";
          const fromStart = fromExpressionElement.contains(range.startContainer) ? getOffsetWithinElement(
            fromExpressionElement,
            range.startContainer,
            range.startOffset
          ) : 0;
          const toEnd = toExpressionElement.contains(range.endContainer) ? getOffsetWithinElement(
            toExpressionElement,
            range.endContainer,
            range.endOffset
          ) : toValue.length;
          dateTimeFromRef.current?.setExpressionValue(
            `${fromValue.slice(0, fromStart)}${char}`
          );
          dateTimeToRef.current?.setExpressionValue(toValue.slice(toEnd));
          requestAnimationFrame(() => {
            focusExpressionInputAtEnd(dateTimeFromRef);
          });
          event.preventDefault();
          event.stopPropagation();
          return;
        }
        dateTimeFromRef.current?.setExpressionValue(char);
        requestAnimationFrame(() => {
          focusExpressionInputAtEnd(dateTimeFromRef);
        });
        event.preventDefault();
        event.stopPropagation();
        return;
      }
      handleSelectionKeyDown(event);
    },
    [
      disabled,
      readOnly,
      handleSelectionKeyDown,
      dateTimeFromRef,
      dateTimeToRef,
      setValue
    ]
  );
  const handleFromTrailingEdge = useCallback(() => {
    dateTimeToRef.current?.focus?.();
  }, []);
  const handleToLeadingEdge = useCallback(() => {
    dateTimeFromRef.current?.focusLast();
  }, [dateTimeFromRef]);
  const handleFromTrailingOverflow = useCallback(
    (overflowChar) => {
      const handlers = Array.from(
        dateTimeToRef.current?.spinButtonHandlers?.values() ?? []
      ).filter((h) => h !== null && h !== void 0);
      const firstHandler = handlers[0];
      if (!firstHandler) {
        return;
      }
      if (value?.to?.type === "expression" || value?.to === null) {
        dateTimeToRef.current?.setExpressionValue(overflowChar);
        requestAnimationFrame(() => {
          const exprEl = dateTimeToRef.current?.expressionInputRef;
          if (exprEl) {
            exprEl.focus();
            const sel = window.getSelection();
            if (sel) {
              const range = document.createRange();
              range.selectNodeContents(exprEl);
              range.collapse(false);
              sel.removeAllRanges();
              sel.addRange(range);
            }
          }
        });
        return;
      }
      if (firstHandler.type === "number") {
        firstHandler.applyText(overflowChar, true);
        firstHandler.element.current?.focus();
        firstHandler.setSelection({ start: 0, end: 0, rtl: true });
      }
    },
    [value?.to]
  );
  return /* @__PURE__ */ jsx(
    Flex,
    {
      ref: rootRef,
      flexDirection: "column",
      "data-testid": dataTestId,
      style: consumerStyle,
      className: consumerClassName,
      ...ariaProps,
      ...behavioralTrackingProps,
      "data-dt-component": COMPONENT_NAME,
      children: /* @__PURE__ */ jsx(FormFieldMessageContext.Provider, { value: formFieldMessageContextValue, children: /* @__PURE__ */ jsxs(InputGroup, { style: { gridTemplateColumns: "1fr auto" }, children: [
        /* @__PURE__ */ jsxs(
          Flex,
          {
            ref: touchedContainerRef,
            className: clsx(
              inputGroupClassName,
              dateTimePicker({
                readOnly,
                showError,
                disabled,
                type: "datetime"
              }),
              textStyle({ textStyle: "base" }),
              fieldTokens({
                color: showError ? "critical" : "neutral",
                disabled: disabledOrAriaDisabled,
                variant: "emphasized"
              }),
              focusClassName,
              styles.dateTimeRange({
                showError,
                disabled: disabledOrAriaDisabled,
                readOnly
              })
            ),
            onCopyCapture: onTimeframeCopyCapture,
            onPasteCapture,
            onBlur: (event) => {
              requestAnimationFrame(() => {
                onBlur?.(event);
              });
            },
            onInput: () => {
              setHasInputContent(
                !!dateTimeFromRef.current?.expressionInputRef?.textContent || !!dateTimeToRef.current?.expressionInputRef?.textContent
              );
            },
            tabIndex: disabled ? void 0 : 0,
            onMouseDown,
            onKeyDownCapture: (e) => {
              onUndoPasteKeyDown(e);
            },
            onKeyDown,
            onFocus: (e) => {
              if (!rootRef.current?.contains(e.relatedTarget)) {
                focusDateTimeInput(e, dateTimeFromRef, dateTimeToRef);
              }
            },
            children: [
              /* @__PURE__ */ jsxs("span", { className: styles.scrollContainer, children: [
                /* @__PURE__ */ jsx(
                  DateTimeInput,
                  {
                    ref: dateTimeFromRef,
                    className: styles.dateTimeInput,
                    disabled,
                    readOnly,
                    placeholder: placeholder.from,
                    name: name === void 0 ? void 0 : `${name}-from`,
                    value: value === null ? null : value.from,
                    onChange: (newValue) => {
                      if (typeof newValue !== "string") {
                        setValue({ from: newValue ?? null });
                      }
                    },
                    "aria-label": dateTimeFromLabel,
                    "aria-disabled": ariaDisabled,
                    required,
                    controlId: id,
                    onTrailingEdge: handleFromTrailingEdge,
                    onTrailingOverflow: handleFromTrailingOverflow,
                    partialTimeAnchorDate: value?.to?.absoluteDate || void 0,
                    onPasteComplete: () => {
                      dateTimeToRef.current?.focus?.();
                    }
                  }
                ),
                /* @__PURE__ */ jsx("span", { className: styles.timeframeSeparator, children: "\u2192" }),
                /* @__PURE__ */ jsx(
                  DateTimeInput,
                  {
                    ref: dateTimeToRef,
                    className: styles.dateTimeInput,
                    disabled,
                    readOnly,
                    placeholder: placeholder.to,
                    name: name === void 0 ? void 0 : `${name}-to`,
                    value: value === null ? null : value.to,
                    onChange: (newValue) => {
                      if (typeof newValue !== "string") {
                        setValue({ to: newValue ?? null });
                      }
                    },
                    "aria-label": dateTimeToLabel,
                    "aria-disabled": ariaDisabled,
                    required,
                    onLeadingEdge: handleToLeadingEdge,
                    dateOnlyPasteAnchor: "endOfDay",
                    partialTimeAnchorDate: value?.from?.absoluteDate || void 0,
                    onPasteComplete: () => {
                      requestAnimationFrame(() => {
                        dateTimeToRef.current?.focusLast();
                      });
                    }
                  }
                )
              ] }),
              !readOnly && !disabledOrAriaDisabled && (value && (value.from?.value !== "" || value.to?.value !== "") || hasInputContent) && /* @__PURE__ */ jsx(
                Button,
                {
                  onFocus: () => {
                    formMessageDispatch?.({
                      type: FormFieldMessagesAction.SETISTOUCHED,
                      payload: { isTouched: true }
                    });
                  },
                  onClick: () => {
                    setValue(null);
                    if (dateTimeFromRef.current?.expressionInputRef) {
                      dateTimeFromRef.current.expressionInputRef.textContent = "";
                    }
                    if (dateTimeToRef.current?.expressionInputRef) {
                      dateTimeToRef.current.expressionInputRef.textContent = "";
                    }
                    setHasInputContent(false);
                    requestAnimationFrame(() => {
                      dateTimeFromRef.current?.expressionInputRef?.focus();
                    });
                  },
                  onKeyDown: (e) => {
                    if (e.key === "Enter") {
                      e.stopPropagation();
                    }
                  },
                  className: clearIconButton,
                  color: showError ? "critical" : "neutral",
                  size: "condensed",
                  "aria-label": clearButtonLabel,
                  children: /* @__PURE__ */ jsx(XmarkIcon, {})
                }
              )
            ]
          }
        ),
        /* @__PURE__ */ jsx(
          DateRangeSelector,
          {
            ref: dateRangeSelectorRef,
            disabled,
            "aria-disabled": ariaDisabled,
            readOnly,
            error: showError,
            "data-dtrum-mask": dataDtrumMask,
            "data-dtrum-allow": dataDtrumAllow
          }
        )
      ] }) })
    }
  );
});
export {
  DateTimeRangePickerRoot
};
