// @dynatrace/strato-components/esm/forms/date-time-range-picker/useSelectionInteractionHandlers.js
import {
  useCallback
} from "react";
import { isExpressionTriggerKey } from "./expression-trigger-helpers.js";
import { buildTimeframeClipboardText } from "./selection-copy-helpers.js";
import {
  extendSelection,
  extendSelectionToAll,
  removeContentEditable,
  getFirstSnapshotHandler,
  findFirstExpressionFocusTarget,
  restoreFocusAfterBulkClear,
  planBulkClear,
  getPartialExpressionElement,
  deletePartialExpression,
  getCrossExpressionElements,
  deleteCrossExpressionPartial,
  consumePreRestoreEndpoints,
  getLastSnapshotHandler
} from "./selection-helpers.js";
function getInputRefs(dateInputFromRef, dateInputToRef) {
  return [dateInputFromRef, ...dateInputToRef ? [dateInputToRef] : []];
}
function useSelectionInteractionHandlers(touchedContainerRef, onFocusInput, dateInputFromRef, dateInputToRef) {
  const onMouseDown = useCallback(
    (event) => {
      if (event.button !== 0) {
        return;
      }
      const root = event.currentTarget;
      const triggerElement = event.target;
      const clickedOnSpinButton = triggerElement instanceof HTMLElement && triggerElement.hasAttribute("contenteditable");
      let initialAnchorSpinButton = null;
      if (!clickedOnSpinButton) {
        const spinButtons = Array.from(
          root.querySelectorAll("[contenteditable]")
        );
        let bestRight = -Infinity;
        for (const sb of spinButtons) {
          const rect = sb.getBoundingClientRect();
          if (rect.right <= event.clientX && rect.right > bestRight) {
            initialAnchorSpinButton = sb;
            bestRight = rect.right;
          }
        }
        if (initialAnchorSpinButton === null && spinButtons.length > 0) {
          initialAnchorSpinButton = spinButtons[0] ?? null;
        }
      }
      window.getSelection()?.removeAllRanges();
      if (root) {
        removeContentEditable(
          root,
          document,
          "mouseup",
          (mouseUpEvent, hasMultiElementSelection) => {
            if (!hasMultiElementSelection && mouseUpEvent.target instanceof HTMLElement && mouseUpEvent.target === triggerElement) {
              if (mouseUpEvent.target.hasAttribute("contenteditable")) {
                mouseUpEvent.target.focus();
              } else {
                onFocusInput();
              }
            }
          },
          initialAnchorSpinButton
        );
      }
    },
    [onFocusInput]
  );
  const onKeyDown = useCallback(
    (event) => {
      const hasModifierKey = event.metaKey === true || event.ctrlKey === true;
      const pickerRoot = event.currentTarget;
      if (event.key === "Shift" || event.key === "Control" || event.key === "Meta" || event.key === "Alt" || event.key === "Command") {
        event.preventDefault();
        return;
      }
      function applyBulkClear(plan, selection) {
        const { snapshot, affectedExpressionRefs, expressionFallbacks } = plan;
        for (const { inputRef, key } of snapshot) {
          inputRef.current?.spinButtonHandlers.get(key)?.clearValue(true);
        }
        for (const [inputRef, text] of expressionFallbacks) {
          inputRef.current?.setExpressionValue(text);
        }
        for (const ref of affectedExpressionRefs) {
          ref.current?.setExpressionValue("");
        }
        selection.removeAllRanges();
      }
      function bulkClear(selection, direction2) {
        const inputRefs = getInputRefs(dateInputFromRef, dateInputToRef);
        const partialExpressionElement = getPartialExpressionElement(
          selection,
          inputRefs
        );
        if (partialExpressionElement) {
          onBeforeExpressionDelete([partialExpressionElement], inputRefs);
          deletePartialExpression(selection, partialExpressionElement);
          return;
        }
        const crossExpressionElements = getCrossExpressionElements(
          selection,
          inputRefs
        );
        if (crossExpressionElements) {
          const isSameExpressionSelection = crossExpressionElements[0] === crossExpressionElements[1];
          const expressionElementsForUndo = isSameExpressionSelection ? [crossExpressionElements[0]] : crossExpressionElements;
          onBeforeExpressionDelete(expressionElementsForUndo, inputRefs);
          const boundaryExpressionRef = isSameExpressionSelection ? inputRefs.find(
            (inputRef) => inputRef.current?.expressionInputRef === crossExpressionElements[0]
          ) : void 0;
          const plan2 = planBulkClear(
            selection,
            inputRefs,
            consumePreRestoreEndpoints(pickerRoot)
          );
          const firstSpinButtonHandler = getFirstSnapshotHandler(plan2.snapshot);
          const expressionFocusTarget2 = findFirstExpressionFocusTarget(
            inputRefs,
            plan2.affectedExpressionRefs,
            plan2.expressionFallbacks
          );
          deleteCrossExpressionPartial(selection, crossExpressionElements);
          for (const { inputRef, key } of plan2.snapshot) {
            inputRef.current?.spinButtonHandlers.get(key)?.clearValue(true);
          }
          for (const [inputRef, text] of plan2.expressionFallbacks) {
            if (inputRef !== boundaryExpressionRef) {
              inputRef.current?.setExpressionValue(text);
            }
          }
          restoreFocusAfterBulkClear(
            firstSpinButtonHandler,
            expressionFocusTarget2
          );
          return;
        }
        const plan = planBulkClear(
          selection,
          inputRefs,
          consumePreRestoreEndpoints(pickerRoot)
        );
        const nextFocusedElementHandler = direction2 === "backward" ? getFirstSnapshotHandler(plan.snapshot) : getLastSnapshotHandler(plan.snapshot);
        const expressionFocusTarget = findFirstExpressionFocusTarget(
          inputRefs,
          plan.affectedExpressionRefs,
          plan.expressionFallbacks
        );
        applyBulkClear(plan, selection);
        restoreFocusAfterBulkClear(
          nextFocusedElementHandler,
          expressionFocusTarget
        );
      }
      function onBeforeExpressionDelete(expressionElements, inputRefs) {
        const expressionSnapshots = [];
        for (const exprEl of expressionElements) {
          let matchingRef;
          for (const inputRef of inputRefs) {
            if (inputRef.current?.expressionInputRef === exprEl) {
              matchingRef = inputRef;
              break;
            }
          }
          if (matchingRef) {
            expressionSnapshots.push({
              inputRef: matchingRef,
              previousText: matchingRef.current?.inputRef?.value ?? exprEl.textContent ?? ""
            });
          }
        }
      }
      if (hasModifierKey && event.key === "a") {
        event.preventDefault();
        if (event.currentTarget instanceof HTMLElement) {
          extendSelectionToAll();
        }
        return;
      }
      if (event.shiftKey === true && (event.key === "ArrowLeft" || event.key === "ArrowRight" || event.key === "Home" || event.key === "End")) {
        event.preventDefault();
        if (event.currentTarget instanceof HTMLElement) {
          extendSelection(event.key);
        }
        return;
      }
      const isCut = hasModifierKey && event.key === "x";
      const isDelete = event.key === "Backspace" || event.key === "Delete";
      const direction = isCut || event.key === "Backspace" ? "backward" : "forward";
      if (isCut || isDelete) {
        const selection = window.getSelection();
        if (selection && !selection.isCollapsed) {
          event.preventDefault();
          const clipboardText = buildTimeframeClipboardText();
          bulkClear(selection, direction);
          if (isCut) {
            if (clipboardText !== null) {
              navigator.clipboard.writeText(clipboardText).catch(() => {
              });
            }
          }
          return;
        }
      }
      if (!hasModifierKey && !event.shiftKey && event.key.length === 1 && (isExpressionTriggerKey(event) || /^\d$/.test(event.key))) {
        const selection = window.getSelection();
        if (selection && !selection.isCollapsed && selection.anchorNode !== selection.focusNode) {
          event.preventDefault();
          const inputRefs = getInputRefs(dateInputFromRef, dateInputToRef);
          const plan = planBulkClear(
            selection,
            inputRefs,
            consumePreRestoreEndpoints(pickerRoot)
          );
          const firstEntry = plan.snapshot[0];
          const firstKey = firstEntry ? [
            ...firstEntry.inputRef.current?.spinButtonHandlers.keys() ?? []
          ][0] : void 0;
          const expressionModeWillSwitch = plan.expressionFallbacks.size > 0 && firstEntry?.key === firstKey;
          if (expressionModeWillSwitch) {
            const firstSpinButtonHandler = getFirstSnapshotHandler(
              plan.snapshot
            );
            const expressionFocusTarget = findFirstExpressionFocusTarget(
              inputRefs,
              plan.affectedExpressionRefs,
              plan.expressionFallbacks
            );
            applyBulkClear(plan, selection);
            expressionFocusTarget?.current?.setExpressionValue(event.key);
            restoreFocusAfterBulkClear(
              firstSpinButtonHandler,
              expressionFocusTarget
            );
            return;
          }
          for (const { inputRef, key } of plan.snapshot) {
            inputRef.current?.spinButtonHandlers.get(key)?.clearValue(true);
          }
          selection.removeAllRanges();
          const handler = firstEntry ? firstEntry.inputRef.current?.spinButtonHandlers.get(
            firstEntry.key
          ) : void 0;
          if (handler) {
            handler.element.current?.focus();
            if (handler.type === "number") {
              handler.applyText(event.key, true);
              const newLength = handler.element.current?.textContent?.length ?? 0;
              handler.setSelection({ start: newLength, end: newLength });
            } else {
              handler.element.current?.dispatchEvent(
                new KeyboardEvent("keydown", {
                  key: event.key,
                  bubbles: true,
                  cancelable: true
                })
              );
            }
          }
          return;
        }
      }
      if (!event.shiftKey && !hasModifierKey && (event.key === "ArrowUp" || event.key === "ArrowDown") && event.target === touchedContainerRef.current) {
        const anchorSpinButton = getAnchorSpinButton();
        if (anchorSpinButton !== null) {
          event.preventDefault();
          anchorSpinButton.focus();
          anchorSpinButton.dispatchEvent(
            new KeyboardEvent("keydown", {
              key: event.key,
              bubbles: true,
              cancelable: true
            })
          );
          return;
        }
      }
      if (event.target === touchedContainerRef.current && !hasModifierKey) {
        const anchorSpinButton = getAnchorSpinButton();
        if (anchorSpinButton !== null) {
          anchorSpinButton.focus();
        } else {
          onFocusInput();
        }
      }
    },
    [onFocusInput, touchedContainerRef, dateInputFromRef, dateInputToRef]
  );
  return { onMouseDown, onKeyDown };
}
function getAnchorSpinButton() {
  const anchorNode = window.getSelection()?.anchorNode ?? null;
  if (anchorNode === null) {
    return null;
  }
  return anchorNode.nodeType === Node.TEXT_NODE ? anchorNode.parentElement?.closest(
    "[contenteditable]"
  ) ?? null : anchorNode?.closest?.("[contenteditable]") ?? null;
}
export {
  useSelectionInteractionHandlers
};
