// @dynatrace/strato-components/charts/core/components/data-point/utils/data-point.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var data_point_exports = {};
__export(data_point_exports, {
  getOuterCircleSize: () => getOuterCircleSize,
  getStrokeColor: () => getStrokeColor
});
module.exports = __toCommonJS(data_point_exports);
var import_lodash_es = require("lodash");
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
var import_constants = require("../../../constants.js");
var import_constants2 = require("../constants.js");
const getStrokeColor = (currentPoints, closestPoint) => {
  if (!closestPoint) {
    return import_colors.default.Text.Neutral.Default;
  }
  return currentPoints.find(
    (currentPoint) => (0, import_lodash_es.isEqual)(currentPoint, closestPoint)
  ) ? import_colors.default.Text.Neutral.Default : import_colors.default.Text.Neutral.Subdued;
};
function getOuterCircleSize(points, valueAxis) {
  const opCoordinateYX = valueAxis === "y" ? "x" : "y";
  const start = points[0][valueAxis];
  const end = points[points.length - 1][valueAxis];
  const minSize = Math.min(start, end);
  const maxSize = Math.max(start, end);
  const baseCoordinate = points[0][opCoordinateYX];
  const coordinateA = baseCoordinate - import_constants.DEFAULT_DATA_POINT_RADIUS - import_constants2.GROUP_PADDING;
  const coordinateB = minSize - import_constants.DEFAULT_DATA_POINT_RADIUS - import_constants2.GROUP_PADDING;
  const groupDiameterWithPadding = import_constants.DEFAULT_DATA_POINT_RADIUS * 2 + import_constants2.GROUP_PADDING * 2;
  const size = maxSize - minSize + groupDiameterWithPadding;
  const diameter = groupDiameterWithPadding;
  if (valueAxis === "y") {
    return {
      x: coordinateA,
      y: coordinateB,
      width: diameter,
      height: size
    };
  }
  return {
    x: coordinateB,
    y: coordinateA,
    width: size,
    height: diameter
  };
}
