// @dynatrace/strato-components/esm/charts/gauge/components/GaugeLabel/GaugeLabel.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import {
  isValidElement
} from "react";
import {
  elementLabelWrapper,
  foreignObjectContainer,
  textContainer
} from "./GaugeLabel.sty.js";
import { GaugeResponsiveLabel } from "../GaugeResponsiveLabel/GaugeResponsiveLabel.js";
const GaugeLabel = ({
  value,
  autoHide,
  responsive,
  prefixIcon,
  ...textProps
}) => {
  const isElement = isValidElement(value);
  if (isElement && responsive) {
    return value;
  }
  if (responsive) {
    return /* @__PURE__ */ jsx(
      GaugeResponsiveLabel,
      {
        text: value?.toString() ?? "",
        id: textProps.id,
        prefixIcon
      }
    );
  }
  if (isElement) {
    return /* @__PURE__ */ jsx(
      "foreignObject",
      {
        x: textProps.x,
        y: textProps.y,
        fontSize: textProps.fontSize,
        className: elementLabelWrapper,
        children: /* @__PURE__ */ jsx(
          "div",
          {
            className: clsx(
              foreignObjectContainer({
                dominantBaseline: textProps.dominantBaseline
              }),
              { [textContainer]: autoHide }
            ),
            children: value
          }
        )
      }
    );
  }
  return /* @__PURE__ */ jsx(
    "text",
    {
      textAnchor: "middle",
      dominantBaseline: "hanging",
      ...textProps,
      className: clsx({ [textContainer]: autoHide }),
      children: value
    }
  );
};
export {
  GaugeLabel
};
