// @dynatrace/strato-components/esm/forms/select/components/Control.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import {
  useMemo,
  Children,
  cloneElement,
  useRef
} from "react";
import { ChevronDownSmallIcon } from "@dynatrace/strato-icons";
import { HiddenSelect } from "./HiddenSelect.js";
import { _isValueContainerComponent } from "./ValueContainer.js";
import { Button } from "../../../buttons/button/Button.js";
import { useFormControlState } from "../../../core/hooks/useFormControlState.js";
import { useId } from "../../../core/hooks/useId.js";
import { useMergeRefs } from "../../../core/hooks/useMergeRefs.js";
import { OverlayStackingProvider } from "../../../core/providers/OverlayStackingProvider.js";
import { useFocusRing } from "../../../core/styles/useFocusRing.js";
import { isVisualRefreshEnabled } from "../../../core/utils/is-visual-refresh-enabled.js";
import { mergeProps as _mergeProps } from "../../../core/utils/merge-props.js";
import { Flex } from "../../../layouts/flex/Flex.js";
import { InternalBaseItem as BaseItem } from "../../forms-core/components/form-field-messages/FormFieldMessageBaseItem.js";
import { useFormFieldMessagesContext } from "../../forms-core/hooks/useFormFieldMessagesContext.js";
import { useSelectMenuContext } from "../hooks/use-select-menu-context.js";
import { SelectActions } from "../state/select-menu-state-reducer.js";
import * as styles from "../styles/Select.sty.js";
function getSelectValueContainer(children) {
  const childrenArray = Children.toArray(children);
  for (const child of childrenArray) {
    if (_isValueContainerComponent(child)) {
      return child;
    }
  }
  throw new Error("Assertion to make sure the container exists.");
}
function Control(props) {
  const { children, isDisabled, innerRef, innerProps, selectProps } = props;
  const ariaLabelledBy = selectProps["aria-labelledby"];
  const {
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    readOnly,
    ariaDisabled,
    ...remainingProps
  } = selectProps.componentProps.trigger ?? {};
  const rootRef = useMergeRefs([
    innerRef,
    selectProps.setTriggerRef
  ]);
  const valueContainer = getSelectValueContainer(children);
  const baseId = useId();
  const triggerId = `strato-select-trigger${baseId}`;
  const customTriggerId = selectProps.customTrigger?.props?.id;
  const { state } = useFormFieldMessagesContext();
  const [hasFormControlError, hint] = useFormControlState(
    selectProps.controlState
  );
  const showError = selectProps.controlState ? hasFormControlError : state.showError;
  const isCritical = isVisualRefreshEnabled() ? showError && !isDisabled : showError;
  const hiddenReactSelectInput = Array.isArray(valueContainer.props.children) ? valueContainer.props.children[1] : null;
  const id = selectProps["selectInputId"];
  const hiddenInput = useMemo(
    () => cloneElement(hiddenReactSelectInput, {
      ...hiddenReactSelectInput.props,
      key: "hidden-react-select-input",
      id,
      autoComplete: "off",
      "aria-haspopup": "listbox",
      "aria-invalid": showError,
      "aria-labelledby": ariaLabelledBy ?? customTriggerId ?? triggerId,
      "aria-label": selectProps["aria-label"],
      "aria-describedby": selectProps["aria-describedby"],
      "aria-details": selectProps["aria-details"]
    }),
    [
      hiddenReactSelectInput,
      id,
      showError,
      ariaLabelledBy,
      customTriggerId,
      triggerId,
      selectProps
    ]
  );
  const { focusProps, focusClassName } = useFocusRing({
    color: showError ? "critical" : "neutral",
    focusWithin: true,
    disabled: isDisabled
  });
  const { dispatch } = useSelectMenuContext();
  const handledByTouchRef = useRef(false);
  const interactionProps = useMemo(() => {
    return {
      onTouchEnd: () => {
        handledByTouchRef.current = true;
        dispatch({
          type: SelectActions.SELECTFOCUS
        });
        requestAnimationFrame(() => {
          selectProps.filterRef?.current?.inputRef?.focus({
            preventScroll: true
          });
        });
      },
      onClick: () => {
        if (handledByTouchRef.current) {
          handledByTouchRef.current = false;
          return;
        }
        dispatch({
          type: SelectActions.SELECTFOCUS
        });
      },
      onKeyDown: (event) => {
        if ((event.key === "Enter" || event.key === " ") && !selectProps.isMulti) {
          dispatch({
            type: SelectActions.CLOSE
          });
        }
      }
    };
  }, [dispatch, selectProps.isMulti]);
  const customTrigger = useMemo(() => {
    if (selectProps.customTrigger) {
      return cloneElement(
        selectProps.customTrigger,
        {
          key: "custom-trigger",
          ref: rootRef,
          id: triggerId,
          ..._mergeProps(
            selectProps.customTrigger.props,
            innerProps,
            ...isDisabled ? [] : [interactionProps]
          )
        },
        [selectProps.customTrigger.props?.children, hiddenInput]
      );
    }
  }, [
    hiddenInput,
    innerProps,
    interactionProps,
    rootRef,
    selectProps.customTrigger,
    triggerId
  ]);
  if (customTrigger) {
    return /* @__PURE__ */ jsx(OverlayStackingProvider, { children: customTrigger });
  }
  return /* @__PURE__ */ jsxs(Flex, { padding: 0, flexDirection: "column", gap: 4, children: [
    /* @__PURE__ */ jsx(OverlayStackingProvider, { children: /* @__PURE__ */ jsxs(
      Button,
      {
        ref: rootRef,
        id: triggerId,
        textAlign: "start",
        "data-testid": dataTestId,
        disabled: isDisabled,
        "aria-disabled": ariaDisabled,
        width: selectProps.triggerWidth,
        ...innerProps,
        variant: "emphasized",
        ...isDisabled ? void 0 : interactionProps,
        color: isCritical ? "critical" : "neutral",
        className: clsx(
          "strato-components-select-control-trigger",
          styles.selectTrigger({
            fullWidth: selectProps.triggerWidth === "full" || selectProps.triggerWidth === "100%",
            showError,
            readOnly
          }),
          focusClassName,
          consumerClassName
        ),
        style: consumerStyle,
        as: "div",
        ...focusProps,
        ...remainingProps,
        children: [
          selectProps.prefix && /* @__PURE__ */ jsx(
            Button.Prefix,
            {
              "data-testid": selectProps.prefix.props["data-testid"],
              className: selectProps.prefix.props.className,
              style: selectProps.prefix.props.style,
              children: selectProps.prefix.props.children
            }
          ),
          /* @__PURE__ */ jsxs(Button.Label, { children: [
            hiddenInput,
            valueContainer
          ] }),
          /* @__PURE__ */ jsx(Button.Suffix, { children: /* @__PURE__ */ jsx(
            "div",
            {
              className: clsx(
                styles.selectTriggerIconWrapper({
                  isOverlayOpen: selectProps.menuIsOpen,
                  readOnly
                })
              ),
              children: /* @__PURE__ */ jsx(ChevronDownSmallIcon, {})
            }
          ) })
        ]
      }
    ) }),
    /* @__PURE__ */ jsx(
      HiddenSelect,
      {
        ref: selectProps.hiddenSelectRef,
        required: selectProps.isRequired,
        selectedValues: selectProps.value,
        name: selectProps.hiddenSelectName,
        disabled: isDisabled || readOnly,
        controlState: selectProps.controlState
      }
    ),
    hint && /* @__PURE__ */ jsx(
      BaseItem,
      {
        variant: showError ? "error" : "hint",
        disabled: isDisabled || ariaDisabled,
        children: hint
      }
    )
  ] });
}
export {
  Control
};
