// @dynatrace/strato-components/esm/forms/select/components/Group.js
import { jsx } from "react/jsx-runtime";
import { useRef, Children } from "react";
import { components } from "react-select";
import {
  applyStoredSelectionSortingOptions,
  sortSelectedOptions
} from "./sort-selected-options-first.js";
import { useBehavioralTrackingProps } from "../../../core/hooks/useBehavioralTrackingProps.js";
function Group(props) {
  const initialRender = useRef(true);
  const sortingMap = useRef(/* @__PURE__ */ new WeakMap());
  const { children, innerProps, ...remainingProps } = props;
  const {
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId
  } = props.data;
  const [groupTrackingProps] = useBehavioralTrackingProps(props.data);
  let childrenArray = Children.toArray(children);
  if (props.selectProps.showSelectedOptionsFirst) {
    if (initialRender.current === true) {
      [childrenArray, sortingMap.current] = sortSelectedOptions(childrenArray);
    } else {
      const [newChildrenArray, newSortingMap] = applyStoredSelectionSortingOptions(childrenArray, sortingMap.current);
      childrenArray = newChildrenArray;
      if (newSortingMap) {
        sortingMap.current = newSortingMap;
      }
    }
  }
  initialRender.current = false;
  return /* @__PURE__ */ jsx(
    components.Group,
    {
      ...remainingProps,
      innerProps: Object.assign(innerProps ?? {}, {
        role: "group",
        className: consumerClassName,
        style: consumerStyle,
        "data-testid": dataTestId,
        ...groupTrackingProps
      }),
      children: childrenArray
    }
  );
}
Group.displayName = "Select.SelectGroup";
export {
  Group
};
