// @dynatrace/strato-components/esm/forms/select/components/GroupHeading.js
import { jsx } from "react/jsx-runtime";
import { useBehavioralTrackingProps } from "../../../core/hooks/useBehavioralTrackingProps.js";
import { Flex } from "../../../layouts/flex/Flex.js";
import { Text } from "../../../typography/text/Text.js";
import * as styles from "../styles/Select.sty.js";
function GroupHeading(props) {
  const {
    id,
    children,
    data: { labelProps }
  } = props;
  const {
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId
  } = labelProps ?? {};
  const [labelTrackingProps] = useBehavioralTrackingProps(labelProps ?? {});
  return /* @__PURE__ */ jsx(
    Flex,
    {
      paddingX: 12,
      paddingTop: 12,
      paddingBottom: 4,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      ...labelTrackingProps,
      children: /* @__PURE__ */ jsx(
        Text,
        {
          as: "span",
          id,
          textStyle: "small-emphasized",
          className: styles.selectGroupHeading,
          children
        }
      )
    }
  );
}
export {
  GroupHeading
};
