// @dynatrace/strato-components/esm/forms/select/components/MenuList.js
import { jsx } from "react/jsx-runtime";
import { useVirtualizer } from "@tanstack/react-virtual";
import clsx from "clsx";
import {
  useEffect,
  useImperativeHandle,
  useRef,
  useState,
  Children
} from "react";
import { NoOptionsMessage } from "./NoOptionsMessage.js";
import {
  applyStoredSelectionSortingOptions,
  sortSelectedOptions
} from "./sort-selected-options-first.js";
import { useMergeRefs } from "../../../core/hooks/useMergeRefs.js";
import * as styles from "../styles/Select.sty.js";
function MenuList(props) {
  const { children, selectProps, innerProps } = props;
  const parentRef = useRef(null);
  const [notifyParentRefChange, setNotifyParentRefChange] = useState();
  const initialRender = useRef(true);
  const sortingMap = useRef(/* @__PURE__ */ new WeakMap());
  let childrenArray = Children.toArray(children);
  const currentChildrenLength = useRef(childrenArray.length);
  if (selectProps.showSelectedOptionsFirst && !(childrenArray.length === 1 && childrenArray[0].type === NoOptionsMessage)) {
    if (initialRender.current === true) {
      [childrenArray, sortingMap.current] = sortSelectedOptions(childrenArray);
    } else if (currentChildrenLength.current !== childrenArray.length) {
      currentChildrenLength.current = childrenArray.length;
      [childrenArray, sortingMap.current] = sortSelectedOptions(childrenArray);
    } else {
      const [newChildrenArray, newSortingMap] = applyStoredSelectionSortingOptions(childrenArray, sortingMap.current);
      childrenArray = newChildrenArray;
      if (newSortingMap) {
        sortingMap.current = newSortingMap;
      }
    }
  }
  initialRender.current = false;
  const optionVirtualizer = useVirtualizer({
    count: childrenArray.length,
    getScrollElement: () => parentRef.current,
    overscan: 5,
    estimateSize: () => 32
  });
  const menulistref = useMergeRefs([
    props.innerRef,
    parentRef,
    setNotifyParentRefChange,
    selectProps.contentRef
  ]);
  useImperativeHandle(
    selectProps.menuListFunctionRef,
    () => ({
      scrollToBottom: () => {
        optionVirtualizer.scrollToIndex(childrenArray.length);
      }
    })
  );
  useEffect(() => {
    requestAnimationFrame(() => {
      optionVirtualizer.measure();
    });
  }, [notifyParentRefChange, childrenArray.length]);
  return /* @__PURE__ */ jsx(
    "div",
    {
      ...innerProps,
      ref: menulistref,
      className: clsx(styles.selectMenuListWrapper, "strato-select-menulist"),
      style: {
        width: optionVirtualizer.scrollOffset !== null && optionVirtualizer.scrollOffset > 0 && selectProps.contentWidth === "max-content" ? optionVirtualizer.scrollRect?.width : void 0
      },
      children: /* @__PURE__ */ jsx(
        "div",
        {
          role: "presentation",
          style: {
            height: optionVirtualizer.getTotalSize(),
            display: "grid",
            gridTemplateRows: optionVirtualizer.measurementsCache.map((el) => `${el.size}px`).join(" ")
          },
          className: styles.selectVirtualizationWrapper,
          children: optionVirtualizer.getVirtualItems().map((virtualRow) => {
            const child = childrenArray?.[virtualRow.index];
            return /* @__PURE__ */ jsx(
              "div",
              {
                className: "strato-select-option",
                "data-index": virtualRow.index,
                style: {
                  gridRow: virtualRow.index + 1,
                  minWidth: 0,
                  height: "fit-content"
                },
                ref: (el) => {
                  optionVirtualizer.measureElement(el);
                },
                children: child
              },
              virtualRow.key
            );
          })
        },
        `wrapper-${selectProps.options.length}`
      )
    }
  );
}
export {
  MenuList
};
