// @dynatrace/strato-components/esm/forms/select/components/Menu.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { assignInlineVars } from "@vanilla-extract/dynamic";
import clsx from "clsx";
import {
  useCallback,
  useEffect,
  useMemo,
  useState
} from "react";
import { createPortal } from "react-dom";
import { FormattedMessage, useIntl } from "react-intl";
import { usePopper } from "react-popper";
import Borders from "@dynatrace/strato-design-tokens/borders";
import { CheckmarkSmallIcon, MinusIcon } from "@dynatrace/strato-icons";
import { Button } from "../../../buttons/button/Button.js";
import { ProgressCircle } from "../../../content/progress/ProgressCircle.js";
import { useId } from "../../../core/hooks/useId.js";
import { useMergeRefs } from "../../../core/hooks/useMergeRefs.js";
import { useOverlayStacking } from "../../../core/hooks/useOverlayStacking.js";
import { OverlayStackingProvider } from "../../../core/providers/OverlayStackingProvider.js";
import { Divider } from "../../../layouts/divider/Divider.js";
import { Flex } from "../../../layouts/flex/Flex.js";
import { Surface } from "../../../layouts/surface/Surface.js";
import { surfaceBorderRadius } from "../../../layouts/surface/variables.sty.js";
import { useOverlay } from "../../../overlays/hooks/useOverlay.js";
import { observeOverlaySize } from "../../../overlays/utils/popper-modifiers.js";
import { Text } from "../../../typography/text/Text.js";
import { _FormFieldContext as FormFieldContext } from "../../forms-core/contexts/FormFieldContext.js";
import { FormFieldMessageContext } from "../../forms-core/contexts/FormFieldMessageContext.js";
import { SearchInput } from "../../search-input/SearchInput.js";
import { useSelectMenuContext } from "../hooks/use-select-menu-context.js";
import { SelectActions } from "../state/select-menu-state-reducer.js";
import * as styles from "../styles/Select.sty.js";
const sameWidthModifier = {
  name: "sameWidth-as-trigger",
  enabled: true,
  fn: ({ state }) => {
    state.styles.popper.width = `${state.rects.reference.width}px`;
    state.styles.popper.height = "min-content";
  },
  phase: "beforeWrite",
  requires: ["computeStyles"]
};
const heightModifier = {
  name: "select-height-handler",
  enabled: true,
  fn: ({ state }) => {
    const viewportHeight = Math.max(
      document.documentElement.clientHeight || 0,
      window.innerHeight || 0
    );
    const topOffset = state.rects.reference.y;
    const bottomOffset = viewportHeight - (state.rects.reference.y + state.rects.reference.height);
    const offset = state.options.placement.includes("bottom") ? bottomOffset : topOffset;
    if (offset < state.rects.popper.height) {
      const maxHeightBase = viewportHeight / 2 > 320 ? viewportHeight / 2 : 320;
      state.styles.popper.maxHeight = `${maxHeightBase - state.rects.reference.height - 20}px`;
    }
  },
  phase: "beforeWrite",
  requires: ["computeStyles"]
};
function Menu(props) {
  const { children, innerRef, innerProps, selectProps } = props;
  const {
    filterRef,
    clearSelectionRef,
    triggerRef,
    inputValue,
    isMulti,
    onInputChange
  } = selectProps;
  const {
    className: contentConsumerClassName,
    style: contentConsumerStyle,
    "data-testid": contentDataTestId,
    ...remainingContentProps
  } = selectProps.componentProps.content ?? {};
  const {
    className: filterConsumerClassName,
    style: filterConsumerStyle,
    "data-testid": filterDataTestId,
    ...remainingFilterProps
  } = selectProps.componentProps.filter ?? {};
  const [portalRef, setPortalRef] = useState(null);
  const [menuWidth, setMenuWidth] = useState(
    triggerRef?.offsetWidth
  );
  const rootPortalRef = useMergeRefs([innerRef, setPortalRef]);
  const stackingContext = useOverlayStacking(void 0, "overlay");
  const { styles: positioningStyles } = usePopper(triggerRef, portalRef, {
    placement: "bottom-start",
    modifiers: [
      {
        name: "offset",
        options: {
          offset: [0, 4]
        }
      },
      sameWidthModifier,
      heightModifier,
      observeOverlaySize
    ]
  });
  const { dispatch } = useSelectMenuContext();
  const clearFilterValue = useCallback(() => {
    if (inputValue !== "") {
      onInputChange("", {
        action: "input-change",
        prevInputValue: inputValue
      });
    }
  }, [inputValue, onInputChange]);
  useOverlay(
    {
      onClose: () => {
        dispatch({
          type: SelectActions.CLOSE
        });
        clearFilterValue();
        setMenuWidth(triggerRef?.offsetWidth);
      },
      isOpen: selectProps.menuIsOpen,
      isDismissible: true,
      shouldCloseOnInteractOutside: (element) => {
        return !selectProps.triggerRef?.contains(element);
      }
    },
    innerRef
  );
  useEffect(() => {
    if (selectProps.menuIsOpen) {
      filterRef?.current?.inputRef?.focus({ preventScroll: true });
    }
  }, [filterRef, selectProps.menuIsOpen]);
  const hiddenTrigger = useMemo(
    () => selectProps.triggerRef?.querySelector("input"),
    [selectProps.triggerRef]
  );
  useEffect(() => {
    return () => {
      if (!document.activeElement || document.activeElement === document.body) {
        hiddenTrigger?.focus();
      }
    };
  }, [hiddenTrigger]);
  const intl = useIntl();
  const ariaLabelId = useId();
  const allSelected = useMemo(() => {
    return isMulti && selectProps.allOptions?.length === selectProps.value?.length;
  }, [isMulti, selectProps.allOptions, selectProps.value]);
  const handleSelectAllInteraction = () => {
    if (!selectProps.loading) {
      if (allSelected) {
        selectProps.onChange([], {
          action: "clear",
          removedValues: selectProps.value
        });
      } else {
        selectProps.onChange(
          selectProps.allOptions,
          {}
        );
      }
    }
  };
  const dataDtProperties = `state:${allSelected ? "deselect" : "select"}`;
  const noOptionsAvailable = selectProps.allOptions === void 0 || selectProps.allOptions.length === 0;
  const isFiltering = selectProps.customFilter && inputValue !== void 0 && inputValue !== "";
  const showClearSelection = selectProps.clearable && selectProps.value && isMulti === false && !isFiltering;
  const showSelectAll = selectProps.clearable && isMulti && !isFiltering && !noOptionsAvailable;
  return createPortal(
    /* @__PURE__ */ jsx(
      Surface,
      {
        "data-testid": contentDataTestId,
        ref: rootPortalRef,
        ...innerProps,
        elevation: "floating",
        padding: 4,
        className: clsx(
          styles.selectMenuOverlay({
            isTriggerWidth: selectProps.contentWidth === "trigger" || selectProps.customFilter && selectProps.contentWidth === "content"
          }),
          contentConsumerClassName
        ),
        style: {
          zIndex: stackingContext.style.zIndex,
          ...positioningStyles.popper,
          width: selectProps.contentWidth !== "trigger" ? selectProps.contentWidth : menuWidth,
          ...contentConsumerStyle,
          ...assignInlineVars({
            [surfaceBorderRadius]: Borders.Radius.Surface.Subdued
          })
        },
        onMouseDown: (e) => e.stopPropagation(),
        ...remainingContentProps,
        children: /* @__PURE__ */ jsxs(OverlayStackingProvider, { children: [
          selectProps.customFilter && // The empty content is need so the TextInput does not consume the required coming from the FormField
          /* @__PURE__ */ jsxs(FormFieldContext.Provider, { value: {}, children: [
            /* @__PURE__ */ jsx(
              FormFieldMessageContext.Provider,
              {
                value: {
                  state: {
                    customMessages: [],
                    formControlMessages: [],
                    showError: false
                  }
                },
                children: /* @__PURE__ */ jsx(
                  SearchInput,
                  {
                    "data-testid": filterDataTestId,
                    ref: filterRef,
                    variant: "minimal",
                    "aria-label": intl.formatMessage({
                      defaultMessage: "Search available options",
                      id: "4PipZ2Z07Q9epib1",
                      description: "Aria label for the filter inside the select"
                    }),
                    placeholder: intl.formatMessage({
                      defaultMessage: "Search",
                      id: "CerADEHqR/6IVmHr",
                      description: "Placeholder for the filter inside the select"
                    }),
                    value: inputValue,
                    onChange: (value) => {
                      onInputChange(value, {
                        action: "input-change",
                        prevInputValue: inputValue
                      });
                    },
                    className: clsx(styles.filter, filterConsumerClassName),
                    style: filterConsumerStyle,
                    ...remainingFilterProps
                  }
                )
              }
            ),
            /* @__PURE__ */ jsx(
              Divider,
              {
                className: clsx(
                  styles.selectMenuDivider,
                  "strato-select-divider"
                )
              }
            )
          ] }),
          showClearSelection && /* @__PURE__ */ jsxs(Fragment, { children: [
            /* @__PURE__ */ jsx(
              Button,
              {
                ref: clearSelectionRef,
                width: "full",
                textAlign: "start",
                onClick: () => {
                  selectProps.onChange(
                    null,
                    {
                      action: "clear",
                      removedValues: selectProps.value
                    }
                  );
                  if (isMulti === false) {
                    dispatch({
                      type: SelectActions.CLOSE
                    });
                    clearFilterValue();
                  }
                },
                onKeyDown: (event) => {
                  if (event.key === " " || event.key === "Enter") {
                    selectProps.onChange(
                      null,
                      {
                        action: "clear",
                        removedValues: selectProps.value
                      }
                    );
                    event.preventDefault();
                    event.stopPropagation();
                    if (isMulti === false) {
                      dispatch({
                        type: SelectActions.CLOSE
                      });
                      clearFilterValue();
                    }
                    if (filterRef?.current && isMulti) {
                      filterRef.current.inputRef?.focus();
                    } else {
                      const hiddentrigger = triggerRef?.querySelector("input");
                      hiddentrigger?.focus();
                    }
                  }
                },
                children: intl.formatMessage({
                  defaultMessage: "Clear selection",
                  id: "/H3rbOvaR6bAasVo",
                  description: "Select component clear selection button."
                })
              }
            ),
            /* @__PURE__ */ jsx(
              Divider,
              {
                className: clsx(
                  styles.selectMenuDivider,
                  "strato-select-divider"
                )
              }
            )
          ] }),
          showSelectAll && /* @__PURE__ */ jsxs(Fragment, { children: [
            /* @__PURE__ */ jsxs(
              Flex,
              {
                tabIndex: 0,
                role: "option",
                className: clsx(
                  "strato-select-option",
                  styles.selectOptionMulti,
                  styles.selectAll,
                  {
                    "strato-select-option--selected": selectProps.value?.length > 0,
                    "strato-select-option--disabled": selectProps.loading
                  }
                ),
                ref: selectProps.clearSelectionRef,
                onClick: () => {
                  handleSelectAllInteraction();
                },
                onKeyDown: (event) => {
                  if (event.key === " " || event.key === "Enter") {
                    handleSelectAllInteraction();
                    event.preventDefault();
                    event.stopPropagation();
                  }
                },
                "data-dt-name": "SelectDeselectAll",
                "data-dt-properties": dataDtProperties,
                children: [
                  /* @__PURE__ */ jsxs(
                    "div",
                    {
                      className: clsx(
                        styles.selectOptionMultiIndicator({
                          selected: selectProps.value?.length > 0
                        })
                      ),
                      children: [
                        allSelected ? /* @__PURE__ */ jsx(
                          CheckmarkSmallIcon,
                          {
                            className: styles.selectOptionMultiIndicatorIcon
                          }
                        ) : null,
                        !allSelected && selectProps.value?.length > 0 ? /* @__PURE__ */ jsx(
                          MinusIcon,
                          {
                            className: styles.selectOptionMultiIndicatorIcon
                          }
                        ) : null
                      ]
                    }
                  ),
                  intl.formatMessage({
                    defaultMessage: "Select all",
                    id: "EsLvnx9UfUL/bZ/r",
                    description: "Select all option in the Select component."
                  })
                ]
              }
            ),
            /* @__PURE__ */ jsx(
              Divider,
              {
                className: clsx(
                  styles.selectMenuDivider,
                  "strato-select-divider"
                )
              }
            )
          ] }),
          selectProps.loading ? /* @__PURE__ */ jsxs(
            Flex,
            {
              paddingY: 6,
              paddingLeft: 8,
              paddingRight: 12,
              className: styles.selectMenuLoadingWrapper,
              children: [
                /* @__PURE__ */ jsx(ProgressCircle, { size: "small", "aria-labelledby": ariaLabelId }),
                /* @__PURE__ */ jsx(Text, { id: ariaLabelId, children: /* @__PURE__ */ jsx(
                  FormattedMessage,
                  {
                    description: "Label for the loading state.",
                    defaultMessage: "Loading",
                    id: "cycz+bYG+VH7Of4e"
                  }
                ) })
              ]
            }
          ) : children
        ] })
      }
    ),
    document.body
  );
}
export {
  Menu
};
