// @dynatrace/strato-components/esm/forms/select/components/NoOptionsMessage.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import { useIntl } from "react-intl";
import { canReceiveDefaultTextFormat as _canReceiveDefaultTextFormat } from "../../../core/utils/can-receive-default-text-format.js";
import { Text } from "../../../typography/text/Text.js";
import * as styles from "../styles/Select.sty.js";
function NoOptionsMessage(props) {
  const selectEmptyState = props.selectProps.emptyState?.props.children;
  const {
    className: consumerClassName,
    style: consumerStyle,
    ...remainingProps
  } = props.selectProps.componentProps.emptyState ?? {};
  const intl = useIntl();
  if (!selectEmptyState || _canReceiveDefaultTextFormat(selectEmptyState)) {
    return /* @__PURE__ */ jsx(
      Text,
      {
        textStyle: "base",
        style: consumerStyle,
        className: clsx(styles.selectNoOptionsMessage, consumerClassName),
        as: "p",
        ...remainingProps,
        children: selectEmptyState ?? intl.formatMessage({
          defaultMessage: "No results found",
          id: "U09aNQxYT1yrJ4eJ",
          description: "Default message when no options are available in the select."
        })
      }
    );
  } else {
    return /* @__PURE__ */ jsx(
      Text,
      {
        textStyle: "base",
        style: consumerStyle,
        className: clsx(styles.selectNoOptionsMessage, consumerClassName),
        as: "p",
        ...remainingProps,
        children: selectEmptyState
      }
    );
  }
}
export {
  NoOptionsMessage
};
