// @dynatrace/strato-components/esm/forms/select/components/Option.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import { CheckmarkSmallIcon } from "@dynatrace/strato-icons";
import { TextEllipsisAroundChildren } from "../../../core/components/TextEllipsisAroundChildren.js";
import { useBehavioralTrackingProps } from "../../../core/hooks/useBehavioralTrackingProps.js";
import { getVirtualFocusStyles } from "../../../core/styles/getVirtualFocusStyles.js";
import { Flex } from "../../../layouts/flex/Flex.js";
import { useSelectMenuContext } from "../hooks/use-select-menu-context.js";
import { SelectActions } from "../state/select-menu-state-reducer.js";
import * as styles from "../styles/Select.sty.js";
function Option(props) {
  const {
    children,
    isSelected,
    innerProps,
    isFocused,
    isDisabled,
    isMulti,
    selectProps
  } = props;
  const {
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "aria-label": ariaLabel,
    prefix,
    suffix
  } = props.data;
  const [behavioralTrackingProps] = useBehavioralTrackingProps(props.data);
  const [prefixTrackingProps] = useBehavioralTrackingProps(
    prefix.componentProps ?? {}
  );
  const [suffixTrackingProps] = useBehavioralTrackingProps(
    suffix.componentProps ?? {}
  );
  const { dispatch } = useSelectMenuContext();
  const isVirtuallyFocused = selectProps.focusOnOption && isFocused;
  const { className: virtualFocusClassName } = getVirtualFocusStyles(
    isVirtuallyFocused,
    isSelected && !isMulti ? "primary" : "neutral"
  );
  return /* @__PURE__ */ jsxs(
    Flex,
    {
      "data-testid": dataTestId,
      role: "option",
      ...innerProps,
      className: clsx(
        "strato-select-option",
        isMulti === true ? styles.selectOptionMulti : styles.selectOptionSingle,
        virtualFocusClassName,
        {
          "strato-select-option--selected": isSelected,
          "strato-select-option--disabled": isDisabled
        },
        consumerClassName
      ),
      style: consumerStyle,
      onClick: (event) => {
        if (!isDisabled) {
          if (isMulti === false) {
            dispatch({
              type: SelectActions.CLOSE
            });
          }
          innerProps.onClick?.(event);
        }
      },
      "aria-selected": isSelected,
      "aria-label": ariaLabel,
      ...behavioralTrackingProps,
      children: [
        isMulti === true ? /* @__PURE__ */ jsx(
          "div",
          {
            className: clsx(
              styles.selectOptionMultiIndicator({
                selected: isSelected
              })
            ),
            children: isSelected === true ? /* @__PURE__ */ jsx(
              CheckmarkSmallIcon,
              {
                className: styles.selectOptionMultiIndicatorIcon
              }
            ) : null
          }
        ) : null,
        props.selectProps.optionsHavePrefix ? /* @__PURE__ */ jsx(
          "span",
          {
            className: clsx(
              styles.selectOptionPrefix,
              prefix.componentProps?.className
            ),
            style: prefix.componentProps?.style,
            "data-testid": prefix.componentProps?.["data-testid"],
            ...prefixTrackingProps,
            children: prefix.component
          }
        ) : null,
        /* @__PURE__ */ jsx(TextEllipsisAroundChildren, { children }),
        suffix.component ? /* @__PURE__ */ jsx(
          "span",
          {
            className: clsx(
              styles.selectOptionSuffix,
              suffix.componentProps?.className
            ),
            style: suffix.componentProps?.style,
            "data-testid": suffix.componentProps?.["data-testid"],
            ...suffixTrackingProps,
            children: suffix.component
          }
        ) : null
      ]
    }
  );
}
export {
  Option
};
