// @dynatrace/strato-components/esm/forms/select/components/Placeholder.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import { useIntl } from "react-intl";
import { components } from "react-select";
import { canReceiveDefaultTextFormat as _canReceiveDefaultTextFormat } from "../../../core/utils/can-receive-default-text-format.js";
import * as styles from "../styles/Select.sty.js";
function Placeholder(props) {
  const { selectProps } = props;
  const intl = useIntl();
  if (selectProps.placeholderText && !_canReceiveDefaultTextFormat(selectProps.placeholderText)) {
    return /* @__PURE__ */ jsx(components.Placeholder, { ...props, className: styles.placeholderColor, children: selectProps.placeholderText });
  }
  return /* @__PURE__ */ jsx(
    components.Placeholder,
    {
      ...props,
      className: clsx(
        styles.placeholderColor,
        styles.selectValueContainerTruncationBox
      ),
      children: selectProps.placeholderText ?? (selectProps.isMulti ? intl.formatMessage({
        defaultMessage: "Select options",
        id: "3ZX0IG+5tekB89nr",
        description: "Default placeholder text for the multi select."
      }) : intl.formatMessage({
        defaultMessage: "Select an option",
        id: "ggfg/kIzwBCNyGCm",
        description: "Default placeholder text for the single select."
      }))
    }
  );
}
export {
  Placeholder
};
