// @dynatrace/strato-components/filters/segment-selector/contexts/SegmentsUI/SegmentsUIReducer.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SegmentsUIReducer_exports = {};
__export(SegmentsUIReducer_exports, {
  segmentsUIReducer: () => segmentsUIReducer
});
module.exports = __toCommonJS(SegmentsUIReducer_exports);
function segmentsUIReducer(state, action) {
  const { type } = action;
  switch (type) {
    case "add": {
      const { segments: segment, isPlaceholder } = action.payload;
      if (segment && !isPlaceholder) {
        return {
          ...state,
          isTouched: true,
          hasSelection: true,
          uiSegments: [...state.uiSegments, { ...segment }],
          focusedSegmentId: segment.id
        };
      }
      const placeholderId = crypto.randomUUID();
      return {
        ...state,
        hasSelection: true,
        uiSegments: [
          ...state.uiSegments,
          {
            id: placeholderId,
            isPlaceholder: true
          }
        ]
      };
    }
    case "remove": {
      const { id } = action.payload;
      const newSegments = state.uiSegments.filter(
        (uiSegment) => id !== uiSegment.id
      );
      const arrayIdentifierErrorIds = state.arrayIdentifierErrorIds?.filter(
        (errorId) => errorId !== id
      );
      if (newSegments.length === 0) {
        return {
          isTouched: true,
          hasSelection: true,
          uiSegments: [
            {
              id: crypto.randomUUID(),
              isPlaceholder: true
            }
          ],
          arrayIdentifierErrorIds: []
        };
      }
      return {
        isTouched: true,
        hasSelection: true,
        uiSegments: newSegments,
        arrayIdentifierErrorIds
      };
    }
    case "removeAll": {
      return {
        isTouched: true,
        hasSelection: false,
        uiSegments: [
          {
            id: crypto.randomUUID(),
            isPlaceholder: true
          }
        ],
        arrayIdentifierErrorIds: []
      };
    }
    case "update": {
      const { newSegment, oldSegment } = action.payload;
      const index = state.uiSegments.findIndex(
        (uiSegment) => uiSegment.id === oldSegment.id
      );
      if (index !== -1) {
        const updatedSegments = [...state.uiSegments];
        updatedSegments[index] = { ...newSegment };
        return {
          ...state,
          isTouched: true,
          hasSelection: state.uiSegments.length > 0,
          uiSegments: updatedSegments,
          focusedSegmentId: newSegment.id
        };
      }
      return state;
    }
    case "clearPlaceholder": {
      const actualSegmentsOnly = state.uiSegments.filter(
        (uiSegment) => !uiSegment.isPlaceholder
      );
      return {
        ...state,
        uiSegments: actualSegmentsOnly
      };
    }
    case "setUISegments": {
      const { payload } = action;
      const { hasSelection, uiSegments } = payload;
      return {
        ...state,
        isTouched: false,
        uiSegments,
        hasSelection: hasSelection ?? state.hasSelection
      };
    }
    case "setArrayIdentifierError": {
      const { payload } = action;
      const { hasError, id } = payload;
      if (hasError) {
        const newState2 = {
          ...state,
          arrayIdentifierErrorIds: state.arrayIdentifierErrorIds ?? []
        };
        newState2.arrayIdentifierErrorIds.push(id);
        return newState2;
      }
      const newState = { ...state };
      newState.arrayIdentifierErrorIds?.filter((errorId) => errorId !== id);
      return newState;
    }
  }
}
