// @dynatrace/strato-components/filters/segment-selector/contexts/SegmentsUI/segmentValidationHelper.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var segmentValidationHelper_exports = {};
__export(segmentValidationHelper_exports, {
  areUiSegmentsValid: () => areUiSegmentsValid,
  getErrorKey: () => getErrorKey,
  validateSegments: () => validateSegments
});
module.exports = __toCommonJS(segmentValidationHelper_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_react_intl = require("react-intl");
var import_formfield_messages_state_reducer = require("../../../../forms/forms-core/state/formfield-messages-state-reducer.js");
var import_segmentMessages = require("../../segmentMessages.js");
var import_SegmentsCache = require("../SegmentsCache/SegmentsCache.js");
async function getErrorKey(uiSegment) {
  if ("isPlaceholder" in uiSegment && uiSegment.isPlaceholder) {
    return null;
  }
  const cachedSegment = await import_SegmentsCache.segmentsClient.queryClient.fetchQuery({
    queryKey: ["getSegment", uiSegment.id],
    queryFn: () => (0, import_SegmentsCache.getSegment)(uiSegment.id).catch(() => null)
  });
  if (!cachedSegment) {
    return "unavailable";
  }
  const uiSegmentHasNoVariables = uiSegment.variables === void 0 || (0, import_lodash_es.isArray)(uiSegment.variables) && uiSegment.variables.length === 0;
  const cachedSegmentHasNoVariables = !cachedSegment.segment.variables;
  if (!cachedSegmentHasNoVariables && uiSegmentHasNoVariables) {
    return "needs-variables";
  }
  if (!uiSegmentHasNoVariables && cachedSegmentHasNoVariables) {
    return "needs-update";
  }
  return null;
}
async function areUiSegmentsValid(uiSegments) {
  if (import_SegmentsCache.segmentsClient.hasAccessRights === false || !uiSegments) {
    return void 0;
  }
  let hasError = false;
  await Promise.all(
    uiSegments?.map(async (uiSegment) => {
      const errorKey = await getErrorKey(uiSegment);
      if (errorKey) {
        hasError = true;
      }
      return errorKey;
    })
  );
  return !hasError;
}
async function validateSegments(segments, onChange, formFieldMessagesDispatch, formControlMessages, errorId, showError = false) {
  onChange();
  const isValid = await areUiSegmentsValid(segments);
  const hasFormControlError = formControlMessages.some(
    (message) => message.id === errorId
  );
  if (hasFormControlError && (import_SegmentsCache.segmentsClient.hasAccessRights === false && segments.length === 0 || isValid)) {
    formFieldMessagesDispatch?.({
      type: import_formfield_messages_state_reducer.FormFieldMessagesAction.REMOVEMESSAGE,
      payload: {
        id: errorId,
        messageType: "formcontrol"
      }
    });
    return;
  }
  if (import_SegmentsCache.segmentsClient.hasAccessRights === false && segments.length > 0) {
    formFieldMessagesDispatch?.({
      type: import_formfield_messages_state_reducer.FormFieldMessagesAction.ADDMESSAGE,
      payload: {
        id: errorId,
        variant: "error",
        validityStateKey: "customError",
        messageType: "formcontrol",
        showError: true,
        message: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_react_intl.FormattedMessage,
          {
            ...import_segmentMessages.SEGMENTSELECTOR_MESSAGES.noReadAndWriteAccess
          }
        )
      }
    });
  }
  if (isValid === false && !hasFormControlError) {
    formFieldMessagesDispatch?.({
      type: import_formfield_messages_state_reducer.FormFieldMessagesAction.ADDMESSAGE,
      payload: {
        id: errorId,
        variant: "error",
        validityStateKey: "customError",
        messageType: "formcontrol",
        showError,
        message: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_intl.FormattedMessage, { ...import_segmentMessages.SEGMENTSELECTOR_MESSAGES.genericSegmentError })
      }
    });
  }
}
