// @dynatrace/strato-components/notifications/notification-settings/components/Content.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Content_exports = {};
__export(Content_exports, {
  Content: () => Content
});
module.exports = __toCommonJS(Content_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_is = require("react-is");
var import_NotifyButtonContext = require("../../../buttons/notify-button/contexts/NotifyButtonContext.js");
var import_can_receive_default_text_format = require("../../../core/utils/can-receive-default-text-format.js");
var import_Flex = require("../../../layouts/flex/Flex.js");
var import_Text = require("../../../typography/text/Text.js");
var import_NotificationSettingsContext = require("../contexts/NotificationSettingsContext.js");
var import_useNotificationMessages = require("../hooks/useNotificationMessages.js");
var styles = __toESM(require("../NotificationSettings.sty.js"));
const Content = (0, import_react.forwardRef)((props, ref) => {
  const { children, ...remainingProps } = props;
  const { overlayAriaLabelledBy } = (0, import_react.useContext)(import_NotifyButtonContext.NotifyButtonContext);
  const { enabled, hasReadPermission, hasWritePermission } = (0, import_react.useContext)(
    import_NotificationSettingsContext.NotificationSettingsContext
  );
  const { headingText, infoText } = (0, import_useNotificationMessages.useNotificationMessages)({
    notificationEnabled: enabled
  });
  const childrenContent = (0, import_react.useMemo)(() => {
    if (typeof children === "function") {
      return children({
        disabled: !enabled,
        headingText,
        infoText,
        hasReadPermission,
        hasWritePermission
      });
    }
    return children;
  }, [
    children,
    enabled,
    hasReadPermission,
    hasWritePermission,
    headingText,
    infoText
  ]);
  if (children === void 0) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { ref, flexDirection: "column", gap: 0, ...remainingProps, children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { id: overlayAriaLabelledBy, textStyle: "base-emphasized", children: headingText }),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { textStyle: "small", className: styles.overlayText, children: infoText })
    ] });
  }
  if ((0, import_can_receive_default_text_format.canReceiveDefaultTextFormat)(childrenContent)) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Text.Text,
      {
        id: overlayAriaLabelledBy,
        ref,
        as: "div",
        textStyle: "base",
        ...remainingProps,
        children: childrenContent
      }
    );
  }
  if ((0, import_react_is.isElement)(childrenContent)) {
    return childrenContent;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { ref, ...remainingProps, children: childrenContent });
});
