// @dynatrace/strato-components/notifications/notification-settings/components/ToggleButton.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ToggleButton_exports = {};
__export(ToggleButton_exports, {
  NotificationToggleButton: () => NotificationToggleButton
});
module.exports = __toCommonJS(ToggleButton_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_Button = require("../../../buttons/button/Button.js");
var import_NotificationSettingsContext = require("../contexts/NotificationSettingsContext.js");
var import_messages = require("../messages.js");
function NotificationToggleButton() {
  const {
    enabled,
    loading,
    toggleNotifications,
    hasReadPermission,
    hasNotFoundError
  } = (0, import_react.useContext)(import_NotificationSettingsContext.NotificationSettingsContext);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Button.Button,
    {
      variant: enabled ? "emphasized" : "accent",
      color: enabled ? "neutral" : "primary",
      onClick: toggleNotifications,
      disabled: !hasReadPermission || hasNotFoundError,
      loading,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_react_intl.FormattedMessage,
        {
          ...import_messages.notificationMessages.notificationsToggle,
          values: { notificationEnabled: enabled }
        }
      )
    }
  );
}
