// @dynatrace/strato-components/filters/filter-field/FilterField.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var FilterField_exports = {};
__export(FilterField_exports, {
  FilterField: () => FilterField
});
module.exports = __toCommonJS(FilterField_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_autocomplete = require("@codemirror/autocomplete");
var import_language = require("@codemirror/language");
var import_state = require("@codemirror/state");
var import_view = require("@codemirror/view");
var import_clsx = __toESM(require("clsx"));
var import_framer_motion = require("framer-motion");
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_react_error_boundary = require("react-error-boundary");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_FilterFieldSuffixButtons = require("./components/FilterFieldSuffixButtons.js");
var import_FilterFieldContext = require("./contexts/FilterFieldContext.js");
var import_FilterFieldLoggerContext = require("./contexts/FilterFieldLoggerContext.js");
var import_edit_mode_update_listener = require("./editor-state-configs/edit-mode-update-listener.js");
var import_editor_config = require("./editor-state-configs/editor-config.js");
var import_pasting_update_listener = require("./editor-state-configs/pasting-update-listener.js");
var import_ErrorFallback = require("./error-fallback/ErrorFallback.js");
var import_filter_value_obfuscation_util = require("./error-fallback/filter-value-obfuscation-util.js");
var import_compartments = require("./extensions/compartments.js");
var import_interactions_extensions = require("./extensions/interactions-extensions.js");
var import_language_extension = require("./extensions/language-extension.js");
var import_recent_saved_filters = require("./extensions/recent-saved-filters.js");
var import_syntax_highlighting_extension = require("./extensions/syntax-highlighting-extension.js");
var styles = __toESM(require("./FilterField.sty.js"));
var import_usePinnedFilterSuggestions = require("./hooks/recent-and-pinned/usePinnedFilterSuggestions.js");
var import_useRecentFilters = require("./hooks/recent-and-pinned/useRecentFilters.js");
var import_useRecentFilterSuggestions = require("./hooks/recent-and-pinned/useRecentFilterSuggestions.js");
var import_useCachedTreeConversion = require("./hooks/useCachedTreeConversion.js");
var import_useErrorAwareViewRef = require("./hooks/useErrorAwareViewRef.js");
var import_useErrorLogsStore = require("./hooks/useErrorLogsStore.js");
var import_useFilterFieldLinting = require("./hooks/useFilterFieldLinting.js");
var import_useSuggestions = require("./hooks/useSuggestions.js");
var import_useUpdatableCallback = require("./hooks/useUpdatableCallback.js");
var import_messages = require("./messages.js");
var import_getNumberParser = require("./parser/getNumberParser.js");
var import_filter_field_state_reducer = require("./state/filter-field-state-reducer.js");
var import_Suggestion = require("./suggestions/Suggestion.js");
var import_SuggestionDetails = require("./suggestions/SuggestionDetails.js");
var import_SuggestionGroup = require("./suggestions/SuggestionGroup.js");
var import_SuggestionGroupLabel = require("./suggestions/SuggestionGroupLabel.js");
var import_suggestions_config = require("./suggestions/suggestions-config.js");
var import_suggestions_extension = require("./suggestions/suggestions-extension.js");
var import_Suggestions = require("./suggestions/Suggestions.js");
var import_SuggestionsOverlay = require("./suggestions/SuggestionsOverlay.js");
var import_auto_format_logical_operators = require("./utils/auto-format-logical-operators.js");
var import_get_next_suggestion = require("./utils/get-next-suggestion.js");
var import_normalize_validator_map = require("./utils/validation-utils/normalize-validator-map.js");
var import_VisuallyHidden = require("../../core/components/visually-hidden/VisuallyHidden.js");
var import_OverlayStacking = require("../../core/contexts/OverlayStacking.js");
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useFormControlState = require("../../core/hooks/useFormControlState.js");
var import_useId = require("../../core/hooks/useId.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_sdk_logger = require("../../core/sdk/sdk-logger.js");
var import_is_slot = require("../../core/utils/is-slot.js");
var editorStyles = __toESM(require("../../editors/base-code-editor/BaseCodeEditor.sty.js"));
var import_FormFieldMessageContextProvider = require("../../forms/forms-core/components/form-field/FormFieldMessageContextProvider.js");
var import_FormFieldMessageBaseItem = require("../../forms/forms-core/components/form-field-messages/FormFieldMessageBaseItem.js");
var import_useFormFieldMessagesContext = require("../../forms/forms-core/hooks/useFormFieldMessagesContext.js");
var import_useFormFieldValidation = require("../../forms/forms-core/hooks/useFormFieldValidation.js");
var import_useSyncedFormControlProps = require("../../forms/forms-core/hooks/useSyncedFormControlProps.js");
var import_formfield_messages_state_reducer = require("../../forms/forms-core/state/formfield-messages-state-reducer.js");
var import_useInputHint = require("../../forms/hooks/useInputHint.js");
var import_Flex = require("../../layouts/flex/Flex.js");
var import_useOverlay = require("../../overlays/hooks/useOverlay.js");
var import_useOverlayWithTrigger = require("../../overlays/hooks/useOverlayWithTrigger.js");
var import_Text = require("../../typography/text/Text.js");
const COMPONENT_NAME = "FilterField";
function isEditorSelection(selection) {
  return selection.ranges !== void 0;
}
const _FilterField = (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const {
      children,
      name,
      value,
      defaultValue,
      required: propRequired,
      disabled: propDisabled,
      readOnly: propReadOnly = false,
      placeholder: placeholderProp,
      id: propId,
      autoSuggestions = false,
      defaultSuggestionsCount: _defaultSuggestionsCount,
      controlState,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "aria-disabled": propAriaDisabled,
      onChange,
      onSuggest,
      onFilter,
      validatorMap,
      parserConfig,
      filterNamespace,
      customTypes,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
    const setCursorPosition = (0, import_useErrorLogsStore.useErrorLogsStore)(
      (state2) => state2.setCursorPosition
    );
    const setAction = (0, import_useErrorLogsStore.useErrorLogsStore)((state2) => state2.setAction);
    const setIsControlled = (0, import_useErrorLogsStore.useErrorLogsStore)((state2) => state2.setIsControlled);
    const suggestionsSlot = import_react.Children.toArray(children).find(
      (child) => (0, import_is_slot.isSlot)(child, import_Suggestions.Suggestions)
    );
    const { getSyntaxTree } = (0, import_useCachedTreeConversion.useCachedTreeConversion)();
    const intl = (0, import_react_intl.useIntl)();
    const { showBoundary } = (0, import_react_error_boundary.useErrorBoundary)();
    const onChangeRef = (0, import_useUpdatableCallback.useUpdatableCallback)(onChange);
    const onFilterRef = (0, import_useUpdatableCallback.useUpdatableCallback)(onFilter);
    const validatorMapRef = (0, import_react.useRef)(validatorMap);
    const customTypesRef = (0, import_react.useRef)(customTypes);
    const internalValidatorMapRef = (0, import_react.useRef)();
    const numberParser = (0, import_getNumberParser.getNumberParser)();
    const isUserEventComposeRef = (0, import_react.useRef)(false);
    const valuePlaceholderWildcardSelection = (0, import_react.useRef)({ isActive: false });
    const {
      id,
      disabled,
      required,
      ariaDisabled,
      readOnly: _readOnly
    } = (0, import_useSyncedFormControlProps.useSyncedFormControlProps)({
      propAriaDisabled,
      propDisabled,
      propId,
      propRequired,
      propReadOnly
    });
    const readOnly = !!_readOnly;
    const [prevValue, setPrevValue] = (0, import_react.useState)(value);
    const [prevPropDisabled, setPrevPropDisabled] = (0, import_react.useState)(disabled);
    const [prevPropRequired, setPrevPropRequired] = (0, import_react.useState)(required);
    const [prevPropReadOnly, setPrevPropReadOnly] = (0, import_react.useState)(readOnly);
    const [prevPropId, setPrevPropId] = (0, import_react.useState)(id);
    const [prevPlaceholderProp, setPrevPlaceholderProp] = (0, import_react.useState)(placeholderProp);
    const [prevOnSuggest, setPrevOnSuggest] = (0, import_react.useState)(() => onSuggest);
    const isControlled = value !== void 0;
    (0, import_react.useEffect)(() => {
      setIsControlled(isControlled);
    }, [isControlled, setIsControlled]);
    const placeholder = intl.formatMessage(import_messages.filterFieldMessages.placeholder);
    const overlayStacking = (0, import_react.useContext)(import_OverlayStacking.OverlayStackingContext);
    const suggestionsOverlayRef = (0, import_react.useRef)(null);
    const {
      targetRef,
      overlayRef: _overlayRef,
      overlayState,
      overlayProps: _overlayProps
    } = (0, import_useOverlayWithTrigger.useOverlayWithTrigger)({
      placement: "bottom-start",
      placementFallbacks: ["top-start"],
      widthStrategy: "trigger",
      offset: 4,
      // Workaround to close all other overlays when opening the suggestions overlay to avoid overlapping overlays.
      // TODO: Refactor with APPDEV-3373.
      onOpenChange(isOpen) {
        if (isOpen && !suggestionsOverlayRef.current && editorViewRef.current?.hasFocus) {
          (0, import_useOverlay.closeAllOverlays)(overlayStacking);
          requestAnimationFrame(() => {
            editorViewRef.current?.focus();
          });
        }
      }
    });
    const overlayRef = (0, import_useMergeRefs.useMergeRefs)([
      suggestionsOverlayRef,
      _overlayRef
    ]);
    const { style: overlayStyles } = _overlayProps;
    const triggerRef = (0, import_react.useRef)(null);
    const editorViewRef = (0, import_react.useRef)(null);
    const internalRootRef = (0, import_react.useRef)(null);
    const rootRef = (0, import_useMergeRefs.useMergeRefs)([
      targetRef,
      internalRootRef
    ]);
    const hiddenInputRef = (0, import_react.useRef)(null);
    const ariaLabelingProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(props);
    const pastingInfoRef = (0, import_react.useRef)({
      pastingInProgress: false
    });
    const { viewDispatch } = (0, import_useErrorAwareViewRef.useErrorAwareViewRef)(editorViewRef);
    const filterFieldLinter = (0, import_useFilterFieldLinting.useFilterFieldLinting)(customTypes);
    const [hasFormControlError, hint] = (0, import_useFormControlState.useFormControlState)(controlState);
    const [hasSyntaxError, setHasSyntaxError] = (0, import_react.useState)(false);
    const {
      state: formFieldMessageState,
      dispatch: formFieldMessageContextDispatch
    } = (0, import_useFormFieldMessagesContext.useFormFieldMessagesContext)();
    const [showError, setShowError] = (0, import_react.useState)(false);
    (0, import_react.useEffect)(() => {
      if (showError && !hasFormControlError && !hasSyntaxError && !formFieldMessageState.showError) {
        setShowError(false);
      }
      if (!editorViewRef.current?.hasFocus && formFieldMessageState.isTouched && (hasFormControlError || formFieldMessageState.showError || hasSyntaxError)) {
        setShowError(true);
      }
    }, [
      hasFormControlError,
      hasSyntaxError,
      formFieldMessageState.showError,
      editorViewRef.current?.hasFocus
    ]);
    const { inputProps, hintProps } = (0, import_useInputHint.useInputHint)({
      hint,
      hasError: showError
    });
    const isDirtyRef = (0, import_react.useRef)(false);
    const [prevDisabled, setPrevDisabled] = (0, import_react.useState)(disabled);
    if (disabled !== prevDisabled) {
      setPrevDisabled(disabled);
      viewDispatch({
        effects: import_editor_config.disabledPropCompartment.reconfigure(
          import_view.EditorView.editable.of(!disabled)
        )
      });
    }
    const [prevReadOnly, setPrevReadOnly] = (0, import_react.useState)(readOnly);
    if (readOnly !== prevReadOnly) {
      setPrevReadOnly(readOnly);
      viewDispatch({
        effects: import_editor_config.readOnlyPropCompartment.reconfigure(
          import_state.EditorState.readOnly.of(readOnly || false)
        )
      });
    }
    const filterFieldCustomErrorRef = (0, import_react.useRef)((0, import_useId.useId)());
    const filterFieldRequiredErrorRef = (0, import_react.useRef)((0, import_useId.useId)());
    const validate = (0, import_react.useCallback)(() => {
      if (valueRef.current === "" && required) {
        formFieldMessageContextDispatch?.({
          type: import_formfield_messages_state_reducer.FormFieldMessagesAction.ADDMESSAGE,
          payload: {
            id: filterFieldRequiredErrorRef.current,
            validityStateKey: "valueMissing",
            variant: "error",
            message: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { textStyle: "small", as: "p", children: (0, import_useFormFieldValidation.getI18nErrorMessage)("valueMissing", hiddenInputRef.current) }),
            messageType: "formcontrol"
          }
        });
      } else if (valueRef.current !== "" && required) {
        formFieldMessageContextDispatch?.({
          type: import_formfield_messages_state_reducer.FormFieldMessagesAction.REMOVEMESSAGE,
          payload: {
            id: filterFieldRequiredErrorRef.current,
            messageType: "formcontrol"
          }
        });
      }
    }, [formFieldMessageContextDispatch, required]);
    const updateSyntaxError = (0, import_react.useCallback)(
      (isValid) => {
        setHasSyntaxError(!isValid);
        if (!isValid) {
          formFieldMessageContextDispatch?.({
            type: import_formfield_messages_state_reducer.FormFieldMessagesAction.ADDMESSAGE,
            payload: {
              id: filterFieldCustomErrorRef.current,
              validityStateKey: "customError",
              variant: "error",
              message: intl.formatMessage(import_messages.filterFieldMessages.syntaxError),
              messageType: "formcontrol"
            }
          });
        } else {
          formFieldMessageContextDispatch?.({
            type: import_formfield_messages_state_reducer.FormFieldMessagesAction.REMOVEMESSAGE,
            payload: {
              id: filterFieldCustomErrorRef.current,
              messageType: "formcontrol"
            }
          });
        }
      },
      [formFieldMessageContextDispatch, setHasSyntaxError, intl]
    );
    const onInputChange = () => {
      if (isDirtyRef.current) {
        validate();
      }
    };
    const onInputBlur = () => {
      isDirtyRef.current = true;
      validate();
      formFieldMessageContextDispatch?.({
        type: import_formfield_messages_state_reducer.FormFieldMessagesAction.SETISTOUCHED,
        payload: { isTouched: true }
      });
    };
    const saveRecentFilter = (0, import_useRecentFilters.useRecentFilterMutation)(filterNamespace);
    (0, import_useRecentFilterSuggestions.useRecentFilterSuggestions)(editorViewRef, filterNamespace);
    (0, import_usePinnedFilterSuggestions.usePinnedFilterSuggestions)(
      editorViewRef,
      filterNamespace,
      overlayState.isOpen
    );
    (0, import_react.useImperativeHandle)(
      forwardedRef,
      () => ({
        element: internalRootRef.current,
        triggerRef: triggerRef.current,
        inputRef: hiddenInputRef.current,
        focus: () => {
          hiddenInputRef.current?.focus();
          editorViewRef.current?.dispatch({
            selection: { anchor: editorViewRef.current.state.doc.length }
          });
        },
        open: () => hiddenInputRef.current?.focus(),
        close: () => overlayState.close(),
        validate: () => {
          validate();
          formFieldMessageContextDispatch?.({
            type: import_formfield_messages_state_reducer.FormFieldMessagesAction.SETISTOUCHED,
            payload: { isTouched: true }
          });
        },
        saveRecentFilter: async () => {
          if (filterNamespace) {
            await saveRecentFilter(valueRef.current).catch(() => void 0);
          }
        }
      }),
      [
        overlayState,
        validate,
        formFieldMessageContextDispatch,
        filterNamespace,
        saveRecentFilter
      ]
    );
    const [state, dispatch] = (0, import_react.useReducer)(import_filter_field_state_reducer.filterFieldReducer, {
      suggestions: /* @__PURE__ */ new Map(),
      activeSuggestion: -1
    });
    const [hideDeleteWhileTyping, setHideDeleteWhileTyping] = (0, import_react.useState)(false);
    const {
      ariaAttributes,
      autoSuggestionsConfig,
      recentSuggestionValues,
      pinnedSuggestionValues,
      logicalOperatorSuggestionValues,
      pastedContentSuggestionValues,
      searchSuggestionValues,
      fallbackKeySuggestionValues,
      statementSuggestionValues,
      focusedListItem,
      suggestionsContext,
      applySuggestion,
      handleNextSuggestion,
      suggestionsOverlayHintId,
      _onSuggest
    } = (0, import_useSuggestions.useSuggestions)({
      ariaDisabled,
      autoSuggestions,
      isOpen: overlayState.isOpen,
      id,
      suggestionsOverlayRef,
      editorViewRef,
      onSuggest,
      viewDispatch,
      hiddenInputRef,
      ariaLabelingProps
    });
    const valueRef = (0, import_react.useRef)(value ?? defaultValue ?? "");
    const transactionRef = (0, import_react.useRef)(void 0);
    const isValidSearchRef = (0, import_react.useRef)(false);
    (0, import_react.useEffect)(() => {
      if (!triggerRef.current) {
        return;
      }
      function handleOnChange(transaction) {
        if (!editorViewRef.current) {
          return transaction;
        }
        let newValue = transaction.newDoc.toString();
        let transactions = [];
        const tree = (0, import_language.syntaxTree)(transaction.state);
        const {
          tree: convertedTree,
          isValid: isValid2,
          isValidSearch: isValidSearch2
        } = getSyntaxTree({
          tree,
          view: editorViewRef.current,
          numberParser,
          doc: transaction.newDoc.toString(),
          internalValidatorMap: internalValidatorMapRef.current,
          customTypes: customTypesRef.current,
          parserConfig
        });
        isValidSearchRef.current = isValidSearch2;
        updateSyntaxError(isValid2);
        onInputChange();
        const annotation = transaction.annotation(import_editor_config.stringAnnotation);
        if (annotation === import_editor_config.ANNOTATION_DOC_CHANGED_CONTROLLED) {
          return transaction;
        }
        if (annotation !== import_editor_config.ANNOTATION_SUGGESTION_APPLY) {
          setAction({
            type: "handleOnChange",
            data: {
              value: valueRef.current,
              newValue
            }
          });
        }
        if (newValue !== valueRef.current) {
          if (isControlled) {
            (0, import_pasting_update_listener.handleControlledPasting)(transaction, pastingInfoRef);
          }
          setHideDeleteWhileTyping(true);
          if (!transaction.isUserEvent("delete.backward") && !transaction.isUserEvent("delete.forward")) {
            [newValue, transactions] = (0, import_auto_format_logical_operators.autoFormatLogicalOperators)(
              newValue,
              tree,
              transaction.selection?.ranges[transaction.selection?.mainIndex ?? 0].to ?? transaction.newDoc.toString().length
            );
          }
          onChangeRef.current?.(newValue, convertedTree, isValid2);
          transactionRef.current = [{ selection: transaction.newSelection }];
          const userEvent = transaction.annotation(import_state.Transaction.userEvent);
          transactionRef.current = [
            {
              ...transaction,
              // On Windows / Linux, hitting 'Shift + Delete' results in an undefined selection,
              // leading to the suggestions callback not being triggered.
              // Even for an empty value, there should always be a selection.
              selection: transaction.newSelection ?? { anchor: 0 },
              // Annotations are not automatically stored, we need to do that manually
              // to be able to determine, which action triggered the transaction when applying
              // the new value in a controlled scenario.
              annotations: import_state.Transaction.userEvent.of(
                userEvent ? userEvent : transaction.annotation(import_editor_config.stringAnnotation) ?? ""
              )
            },
            ...transactions
          ];
          if (!isControlled) {
            valueRef.current = newValue;
            onInputChange();
          }
          return isControlled ? [] : [transaction, ...transactions];
        }
        return [transaction];
      }
      internalValidatorMapRef.current = (0, import_normalize_validator_map.toInternalValidatorMap)(validatorMap);
      const initialEditorState = import_state.EditorState.create({
        doc: value ?? defaultValue ?? "",
        extensions: [
          import_language_extension.FilterFieldLanguageExtension,
          filterFieldLinter,
          import_syntax_highlighting_extension.FilterFieldSyntaxHighlightingExtension,
          (0, import_interactions_extensions.FilterFieldInteractionsExtension)({
            closeSuggestions: overlayState.close,
            showBoundary
          }),
          (0, import_compartments.DynamicFilterFieldConfigurations)({
            ariaAttributes,
            disabled: !!disabled,
            readOnly,
            placeholder: placeholderProp ?? placeholder
          }),
          import_view.EditorView.domEventObservers({
            compositionend: () => {
              isUserEventComposeRef.current = false;
              editorViewRef.current?.dispatch({
                annotations: import_editor_config.stringAnnotation.of(import_editor_config.ANNOTATION_COMPOSE_END)
              });
            }
          }),
          import_recent_saved_filters.FilterFieldRecentSavedFiltersExtension,
          /**
           * Suggestions
           */
          import_editor_config.suggestionsCompartment.of(
            (0, import_suggestions_extension.suggestionsExtension)({
              overlayState,
              numberParser,
              onSuggest,
              internalValidatorMap: internalValidatorMapRef.current,
              _onSuggest,
              getSyntaxTree,
              parserConfig,
              customTypes: customTypesRef.current
            })
          ),
          /**
           * Listeners
           */
          // Update cursor position for error logging whenever it changes.
          import_state.EditorState.transactionFilter.of((transaction) => {
            const userEvent = transaction.annotation(import_state.Transaction.userEvent);
            if (userEvent?.includes("select")) {
              setCursorPosition(transaction.state.selection.main.anchor);
            }
            return transaction;
          }),
          // Intercept transactions that change the doc to not instantly apply the changes
          // in a controlled scenario. Instead, the transaction is stored so we can apply
          // it if the consumer actually updates the value.
          import_state.EditorState.transactionFilter.of((transaction) => {
            if (!editorViewRef.current) {
              return transaction;
            }
            const annotation = transaction.annotation(import_editor_config.stringAnnotation);
            if (annotation === import_editor_config.ANNOTATION_DOC_CHANGED_CONTROLLED) {
              return handleOnChange(transaction);
            }
            if (annotation === import_editor_config.ANNOTATION_END_PASTING_MODE) {
              queueMicrotask(() => {
                try {
                  if (editorViewRef?.current) {
                    (0, import_autocomplete.startCompletion)(editorViewRef.current);
                  }
                } catch (e) {
                  showBoundary(e);
                }
              });
            }
            let userEvent = transaction.annotation(import_state.Transaction.userEvent);
            if (userEvent === void 0) {
              switch (annotation) {
                case import_editor_config.ANNOTATION_COMPOSE_END:
                  userEvent = import_editor_config.ANNOTATION_COMPOSE_END;
                  break;
                case import_editor_config.ANNOTATION_DELETE_STATEMENT:
                  userEvent = import_editor_config.ANNOTATION_DELETE_STATEMENT;
                  break;
                case import_editor_config.ANNOTATION_SUGGESTION_APPLY: {
                  userEvent = import_editor_config.ANNOTATION_SUGGESTION_APPLY;
                  const selection = transaction.selection?.ranges;
                  if (selection && selection[0] && selection[0].to - selection[0].from === 5) {
                    valuePlaceholderWildcardSelection.current = {
                      isActive: true,
                      range: { from: selection[0].from, to: selection[0].to },
                      potentialNewValue: transaction.newDoc.toString()
                    };
                  }
                  break;
                }
              }
            }
            if (userEvent === "input.type.compose.start") {
              isUserEventComposeRef.current = true;
            }
            if (!userEvent || isUserEventComposeRef.current) {
              return transaction;
            }
            if ((transaction.docChanged && userEvent || !isUserEventComposeRef.current) && // We add the space character in the updateListener, so we don't want to call OnChange.
            !transaction.isUserEvent("select")) {
              return handleOnChange(transaction);
            }
            return transaction;
          }),
          import_view.EditorView.updateListener.of((update) => {
            const isUndoOrRedo = update.transactions.some(
              (transaction) => transaction.isUserEvent("undo") || transaction.isUserEvent("redo")
            );
            if (update.docChanged && isUndoOrRedo) {
              update.transactions.forEach(
                (transaction) => handleOnChange(transaction)
              );
              editorView?.dispatch({
                effects: [
                  import_editor_config.pastingEffect.of({
                    pastingInProgress: false
                  })
                ]
              });
            }
          }),
          // Update Listener for pasting
          import_view.EditorView.updateListener.of((update) => {
            (0, import_pasting_update_listener.handlePastingTransaction)(
              update,
              pastingInfoRef,
              editorView,
              editorViewRef
            );
          }),
          // Automatically insert a space at the end of the input when entering edit mode,
          // if the cursor is at the end, the statement is complete, and no trailing space or bracket is present.
          // This starts a new statement.
          import_view.EditorView.updateListener.of((update) => {
            (0, import_edit_mode_update_listener.handleEditMode)(update, suggestionsContext);
          })
        ]
      });
      const editorView = new import_view.EditorView({
        state: initialEditorState,
        parent: triggerRef.current
      });
      editorViewRef.current = editorView;
      if (validatorMap) {
        editorView.dispatch({
          effects: [import_editor_config.validatorMapEffect.of(validatorMapRef?.current)],
          // In order to re-run the decorations and re-generate errors and diagnostics,
          // we have to dispatch a new document.
          changes: {
            from: 0,
            to: editorViewRef.current.state.doc.length,
            insert: editorViewRef.current.state.doc
          }
        });
      }
      const { isValid, isValidSearch } = getSyntaxTree({
        tree: (0, import_language.syntaxTree)(editorView.state),
        view: editorView,
        numberParser,
        doc: editorView.state.doc.toString(),
        internalValidatorMap: internalValidatorMapRef.current,
        customTypes: customTypesRef.current,
        parserConfig
      });
      isValidSearchRef.current = isValidSearch;
      updateSyntaxError(isValid);
      return () => editorView.destroy();
    }, []);
    const validatorMapChanged = !(0, import_lodash_es.isEqual)(validatorMap, validatorMapRef.current);
    const customTypesChanged = !(0, import_lodash_es.isEqual)(customTypes, customTypesRef.current);
    if ((validatorMapChanged || customTypesChanged) && editorViewRef.current) {
      if (validatorMapChanged) {
        validatorMapRef.current = validatorMap;
        internalValidatorMapRef.current = validatorMap ? (0, import_normalize_validator_map.toInternalValidatorMap)(validatorMap) : void 0;
      }
      if (customTypesChanged) {
        customTypesRef.current = customTypes;
      }
      if (validatorMapChanged || customTypesChanged) {
        viewDispatch({
          effects: import_editor_config.suggestionsCompartment.reconfigure(
            (0, import_suggestions_extension.suggestionsExtension)({
              overlayState,
              numberParser,
              onSuggest,
              _onSuggest,
              getSyntaxTree,
              internalValidatorMap: internalValidatorMapRef.current,
              customTypes: customTypesRef.current,
              parserConfig
            })
          )
        });
      }
      const doc = editorViewRef.current.state.doc;
      queueMicrotask(() => {
        try {
          if (editorViewRef.current) {
            const tree = (0, import_language.syntaxTree)(editorViewRef.current.state);
            const { tree: convertedTree, isValid } = getSyntaxTree({
              tree,
              view: editorViewRef.current,
              numberParser,
              doc: doc.toString(),
              internalValidatorMap: internalValidatorMapRef.current,
              customTypes: customTypesRef.current,
              parserConfig
            });
            updateSyntaxError(isValid);
            onChangeRef.current?.(doc.toString(), convertedTree, isValid);
            onInputChange();
            viewDispatch({
              effects: [import_editor_config.validatorMapEffect.of(validatorMapRef?.current)],
              // In order to re-run the decorations and re-generate errors and diagnostics,
              // we have to dispatch a new document.
              changes: {
                from: 0,
                to: editorViewRef.current.state.doc.length,
                insert: doc.toString()
              }
            });
            if (validatorMapChanged || customTypesChanged) {
              queueMicrotask(() => {
                try {
                  if (editorViewRef.current) {
                    (0, import_autocomplete.startCompletion)(editorViewRef.current);
                  }
                } catch (e) {
                  showBoundary(e);
                }
              });
            }
          }
        } catch (e) {
          showBoundary(e);
        }
      });
    }
    if (placeholderProp !== prevPlaceholderProp) {
      viewDispatch({
        effects: import_editor_config.placeholderCompartment.reconfigure(
          (0, import_view.placeholder)(placeholderProp ?? placeholder)
        )
      });
      setPrevPlaceholderProp(placeholderProp);
    }
    if (onSuggest !== prevOnSuggest) {
      viewDispatch({
        effects: import_editor_config.suggestionsCompartment.reconfigure(
          (0, import_suggestions_extension.suggestionsExtension)({
            overlayState,
            numberParser,
            onSuggest,
            _onSuggest,
            getSyntaxTree,
            internalValidatorMap: internalValidatorMapRef.current,
            customTypes: customTypesRef.current,
            parserConfig
          })
        )
      });
      setPrevOnSuggest(() => onSuggest);
    }
    if (value !== prevValue && isControlled) {
      setPrevValue(value);
      valueRef.current = value;
      queueMicrotask(() => {
        try {
          setAction({
            type: "controlledValueChange",
            data: {
              value: prevValue ?? "",
              newValue: valueRef.current
            }
          });
          onInputChange();
          const lastTransaction = transactionRef.current && transactionRef.current[transactionRef.current.length - 1];
          const selection = lastTransaction?.selection ?? // Fall back to the current editor selection in case the last transaction is undefined
          // to avoid a jumping cursor when typing really fast.
          editorViewRef.current?.state.selection;
          const anchor = Math.min(
            value.length,
            (isEditorSelection(selection) ? selection.ranges[0].anchor : selection?.anchor) ?? value.length
          );
          let dispatchedSelection = {
            anchor
          };
          if (valuePlaceholderWildcardSelection.current?.isActive && valuePlaceholderWildcardSelection.current?.potentialNewValue === value) {
            const selectionFrom = valuePlaceholderWildcardSelection.current.range?.from;
            const selectionTo = valuePlaceholderWildcardSelection.current.range?.to;
            if (selectionFrom && selectionFrom <= value.length && selectionTo && selectionTo <= value.length) {
              dispatchedSelection = {
                anchor: selectionFrom,
                head: selectionTo
              };
            }
          }
          viewDispatch({
            changes: {
              from: 0,
              to: editorViewRef.current?.state.doc.length ?? 0,
              insert: value
            },
            annotations: import_editor_config.stringAnnotation.of(import_editor_config.ANNOTATION_DOC_CHANGED_CONTROLLED),
            selection: dispatchedSelection
          });
          if (valuePlaceholderWildcardSelection.current) {
            valuePlaceholderWildcardSelection.current = {
              isActive: false
            };
          }
          transactionRef.current = void 0;
        } catch (e) {
          showBoundary(e);
        }
      });
    }
    if (disabled !== prevPropDisabled || required !== prevPropRequired || readOnly !== prevPropReadOnly || id !== prevPropId) {
      setPrevPropDisabled(disabled);
      setPrevPropRequired(required);
      setPrevPropReadOnly(readOnly);
      setPrevPropId(id);
      if (disabled !== prevPropDisabled) {
        viewDispatch({
          effects: import_editor_config.disabledPropCompartment.reconfigure(
            import_view.EditorView.editable.of(!disabled)
          )
        });
      }
      if (readOnly !== prevPropReadOnly) {
        viewDispatch({
          effects: import_editor_config.readOnlyPropCompartment.reconfigure(
            import_state.EditorState.readOnly.of(readOnly)
          )
        });
      }
    }
    const [hasAriaLivePolite, setHasAriaLivePolite] = (0, import_react.useState)(false);
    (0, import_react.useEffect)(() => {
      setHasAriaLivePolite(false);
      const debounceTimer = setTimeout(() => {
        setHasAriaLivePolite(true);
      }, 300);
      return () => {
        clearTimeout(debounceTimer);
      };
    }, [valueRef.current]);
    const defaultSuggestionsCount = (0, import_react.useMemo)(() => {
      if (typeof _defaultSuggestionsCount === "number") {
        return _defaultSuggestionsCount;
      }
      return {
        empty: _defaultSuggestionsCount?.empty ?? import_suggestions_config.DEFAULT_SUGGESTIONS_COUNT.empty,
        filled: _defaultSuggestionsCount?.filled ?? import_suggestions_config.DEFAULT_SUGGESTIONS_COUNT.filled
      };
    }, [_defaultSuggestionsCount]);
    const filterFieldContextValue = (0, import_react.useMemo)(
      () => ({
        hasAriaLivePolite,
        applySuggestion,
        dispatch,
        focusedListItem,
        filterFieldId: id,
        term: suggestionsContext.current?.token.textValue,
        autoSuggestions: autoSuggestions ? autoSuggestionsConfig : {},
        recentSuggestionValues,
        pinnedSuggestionValues,
        logicalOperatorSuggestionValues,
        pastedContentSuggestionValues,
        searchSuggestionValues,
        fallbackKeySuggestionValues,
        fallbackKeys: internalValidatorMapRef.current?.fallbackKeys,
        statementSuggestionValues,
        filterStatementKeysMap: internalValidatorMapRef.current?.filterStatementKeysMap,
        triggerRef: triggerRef.current ?? null,
        suggestionsOverlayRef: suggestionsOverlayRef.current ?? null,
        filterNamespace: filterNamespace ?? null,
        editorView: editorViewRef.current,
        parserConfig,
        isValidSearch: isValidSearchRef,
        defaultSuggestionsCount,
        handleNextSuggestion
      }),
      [
        hasAriaLivePolite,
        applySuggestion,
        focusedListItem,
        id,
        suggestionsContext,
        autoSuggestions,
        autoSuggestionsConfig,
        recentSuggestionValues,
        pinnedSuggestionValues,
        logicalOperatorSuggestionValues,
        pastedContentSuggestionValues,
        searchSuggestionValues,
        fallbackKeySuggestionValues,
        internalValidatorMapRef,
        statementSuggestionValues,
        filterNamespace,
        parserConfig,
        defaultSuggestionsCount,
        handleNextSuggestion
      ]
    );
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_FilterFieldContext.FilterFieldContext.Provider, { value: filterFieldContextValue, children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      import_Flex.Flex,
      {
        ref: rootRef,
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        gap: 4,
        flexDirection: "column",
        className: (0, import_clsx.default)(styles.wrapper, consumerClassName),
        style: consumerStyle,
        onMouseMove: () => {
          setHideDeleteWhileTyping(false);
        },
        ...behavioralTrackingProps,
        "data-dt-component": COMPONENT_NAME,
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
            "div",
            {
              ...inputProps,
              "aria-invalid": controlState ? inputProps["aria-invalid"] : hasSyntaxError,
              className: (0, import_clsx.default)(
                editorStyles.editorContainer({
                  size: "default",
                  fullHeight: false,
                  // check if the editor is disabled first to apply proper styling
                  hasError: !(disabled || ariaDisabled) && showError,
                  lineNumbers: "hidden",
                  readOnly: disabled || readOnly
                }),
                styles.container,
                props.className,
                {
                  [styles.errorText]: showError,
                  [styles.errorVars]: showError,
                  [styles.disabled]: disabled,
                  [styles.readOnly]: readOnly,
                  [styles.ariaDisabled]: ariaDisabled,
                  hideDeleteWhileTyping
                }
              ),
              onBlur: onInputBlur,
              onKeyDown: (event) => {
                if (event.target instanceof HTMLElement && event.target.closest("button")) {
                  return;
                }
                if (readOnly && event.key !== "Tab" && !event.ctrlKey) {
                  event.preventDefault();
                  return;
                }
                switch (event.key) {
                  case "Space":
                    if (!editorViewRef.current) {
                      return;
                    }
                    if (event.ctrlKey || event.altKey) {
                      overlayState.open();
                      (0, import_autocomplete.startCompletion)(editorViewRef.current);
                    }
                    break;
                  case "Enter": {
                    event.preventDefault();
                    if (event.shiftKey || !editorViewRef.current) {
                      return;
                    }
                    if (!focusedListItem || event.ctrlKey || event.metaKey) {
                      validate();
                      const { tree, isValid } = getSyntaxTree({
                        tree: (0, import_language.syntaxTree)(editorViewRef.current.state),
                        view: editorViewRef.current,
                        numberParser,
                        doc: editorViewRef.current.state.doc.toString(),
                        internalValidatorMap: internalValidatorMapRef.current,
                        customTypes,
                        parserConfig
                      });
                      onFilterRef.current?.({
                        value: valueRef.current,
                        syntaxTree: tree,
                        isValid
                      });
                      overlayState.close();
                      if (isValid && hiddenInputRef.current?.form) {
                        let inputCount = 0;
                        for (const element of Array.from(
                          hiddenInputRef.current.form.elements
                        )) {
                          if (!element.hasAttribute("hidden") && (element.tagName === "INPUT" && element.getAttribute("type") !== "hidden" || element.tagName === "SELECT" || element.tagName === "TEXTAREA")) {
                            inputCount += 1;
                            if (inputCount > 1) {
                              break;
                            }
                          }
                        }
                        if (inputCount === 1) {
                          hiddenInputRef.current?.form?.requestSubmit();
                        }
                      } else {
                        setTimeout(() => {
                          hiddenInputRef.current?.form?.reportValidity();
                        });
                      }
                    } else if (focusedListItem && focusedListItem.id) {
                      if (state.suggestions.has(focusedListItem.id)) {
                        state.suggestions.get(focusedListItem.id)?.();
                      } else {
                        const focusedElement = document.querySelector(
                          `[data-strato-filter-field-virtual-focus-id="${focusedListItem?.id}"]`
                        );
                        focusedElement?.click();
                      }
                      return;
                    }
                    break;
                  }
                }
                if (overlayState.isOpen) {
                  if (import_editor_config.overrideKeys.includes(event.key)) {
                    event.preventDefault();
                    event.stopPropagation();
                  }
                  if (event.key === "ArrowUp" || event.key === "ArrowDown") {
                    const direction = event.key === "ArrowUp" ? "previous" : "next";
                    const overlay = suggestionsOverlayRef.current;
                    const currentlyFocusedItem = (0, import_get_next_suggestion.getCurrentFocusableListItem)(
                      overlay,
                      focusedListItem?.id
                    );
                    const isAtEnd = currentlyFocusedItem === (0, import_get_next_suggestion.getLastFocusableListItem)(overlay) && direction === "next";
                    const isAtStart = currentlyFocusedItem === (0, import_get_next_suggestion.getFirstFocusableListItem)(overlay) && direction === "previous";
                    if (isAtEnd || isAtStart) {
                      const scrollContainer = overlay?.querySelector(
                        '[role="listbox"]'
                      )?.firstElementChild;
                      if (scrollContainer) {
                        scrollContainer.scroll({
                          top: direction === "next" ? 0 : scrollContainer.scrollHeight
                        });
                        requestAnimationFrame(() => {
                          const nextSuggestion2 = (0, import_get_next_suggestion.getNextFocusableListItem)(
                            direction,
                            overlay,
                            focusedListItem?.id
                          );
                          handleNextSuggestion(nextSuggestion2);
                        });
                      }
                      return;
                    }
                    const nextSuggestion = (0, import_get_next_suggestion.getNextFocusableListItem)(
                      direction,
                      overlay,
                      focusedListItem?.id
                    );
                    handleNextSuggestion(nextSuggestion);
                  }
                }
              },
              onKeyDownCapture: (event) => {
                if ((event.key === "Home" || event.key === "End" || event.key === "PageUp" || event.key === "PageDown") && focusedListItem) {
                  event.stopPropagation();
                  event.preventDefault();
                  let nextSuggestion = null;
                  if (event.key === "PageUp" || event.key === "PageDown") {
                    const direction = event.key === "PageUp" ? "previous" : "next";
                    nextSuggestion = (0, import_get_next_suggestion.getNextFocusableListItem)(
                      direction,
                      suggestionsOverlayRef.current,
                      focusedListItem?.id,
                      5
                    );
                    handleNextSuggestion(nextSuggestion);
                  } else if (event.key === "Home" || event.key === "End") {
                    const overlay = suggestionsOverlayRef.current;
                    const scrollContainer = overlay?.querySelector(
                      '[role="listbox"]'
                    )?.firstElementChild;
                    if (scrollContainer) {
                      scrollContainer.scroll({
                        top: event.key === "Home" ? 0 : scrollContainer.scrollHeight
                      });
                    }
                    requestAnimationFrame(() => {
                      const nextSuggestion2 = event.key === "Home" ? (0, import_get_next_suggestion.getFirstFocusableListItem)(overlay) : (0, import_get_next_suggestion.getLastFocusableListItem)(overlay);
                      handleNextSuggestion(nextSuggestion2);
                    });
                  }
                }
              },
              onMouseDown: (event) => {
                if (disabled) {
                  event.preventDefault();
                }
              },
              children: [
                /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  "span",
                  {
                    className: styles.filterIcon({
                      hasError: showError,
                      disabled: disabled || Boolean(ariaDisabled)
                    }),
                    children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.FilterIcon, {})
                  }
                ),
                /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { ref: triggerRef, style: { maxHeight: "inherit" } }),
                /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  import_FilterFieldSuffixButtons.FilterFieldSuffixButtons,
                  {
                    value: valueRef.current,
                    ariaDisabled,
                    disabled,
                    showError,
                    readOnly
                  }
                )
              ]
            }
          ),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_framer_motion.AnimatePresence, { children: !readOnly && overlayState.isOpen && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_SuggestionsOverlay.SuggestionsOverlay,
            {
              ref: overlayRef,
              "data-testid": dataTestId ? `filterfield-suggestions-${dataTestId}` : void 0,
              overlayProps: { style: overlayStyles },
              "data-dtrum-mask": dataDtrumMask,
              "data-dtrum-allow": dataDtrumAllow,
              children: suggestionsSlot === void 0 ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Suggestions.Suggestions, { children }) : children
            }
          ) }),
          hint && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_FormFieldMessageBaseItem.InternalBaseItem,
            {
              variant: showError ? "error" : "hint",
              disabled: disabled || ariaDisabled,
              ...hintProps,
              children: hint
            }
          ),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_VisuallyHidden.VisuallyHidden, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "input",
            {
              ref: hiddenInputRef,
              name,
              value: valueRef.current,
              required,
              disabled,
              readOnly,
              tabIndex: -1,
              id,
              onFocus: () => {
                editorViewRef.current?.focus();
              },
              onChange: () => {
              },
              onInvalid: (event) => {
                event.preventDefault();
                setShowError(true);
                event.currentTarget.focus();
                onInputBlur();
              },
              "aria-hidden": "true"
            }
          ) }),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "p",
            {
              id: suggestionsOverlayHintId,
              className: styles.suggestionsOverlayAnnouncement,
              "aria-hidden": "true",
              children: intl.formatMessage(
                import_messages.filterFieldMessages.suggestions.a11y.overlayAnnouncement
              )
            }
          )
        ]
      }
    ) });
  }
);
_FilterField.displayName = COMPONENT_NAME;
const InternalFilterField = (0, import_FormFieldMessageContextProvider.withFormFieldMessageContextProvider)(
  (0, import_react.forwardRef)((props, forwardedRef) => {
    const stratoLogger = (0, import_react.useMemo)(
      () => import_sdk_logger.createStratoLogger.bind({
        name: props.name,
        id: props.id
      })(COMPONENT_NAME),
      [props.name, props.id]
    );
    const cursorPosition = (0, import_useErrorLogsStore.useErrorLogsStore)((state) => state.cursorPosition);
    const action = (0, import_useErrorLogsStore.useErrorLogsStore)((state) => state.action);
    const isControlled = (0, import_useErrorLogsStore.useErrorLogsStore)((state) => state.isControlled);
    const loggerContext = (0, import_react.useMemo)(
      () => ({ logger: stratoLogger }),
      [stratoLogger]
    );
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_react_error_boundary.ErrorBoundary,
      {
        onError: (error) => {
          if (action !== null) {
            action.data.value = (0, import_filter_value_obfuscation_util.obfuscateFilterString)(
              action.data.value,
              props.validatorMap,
              props.parserConfig
            );
            if (action.type === "handleOnChange" || action.type === "controlledValueChange") {
              action.data.newValue = (0, import_filter_value_obfuscation_util.obfuscateFilterString)(
                action.data.newValue,
                props.validatorMap,
                props.parserConfig
              );
            }
          }
          stratoLogger.error(error?.message, error, {
            filterString: (0, import_filter_value_obfuscation_util.obfuscateFilterString)(
              props.value ?? props.defaultValue ?? "",
              props.validatorMap,
              props.parserConfig
            ),
            cursorPosition,
            action,
            isControlled
          });
        },
        FallbackComponent: (fallbackProps) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ErrorFallback.ErrorFallback, { ...fallbackProps }),
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_FilterFieldLoggerContext.FilterFieldLoggerContext.Provider, { value: loggerContext, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(_FilterField, { ...props, ref: forwardedRef }) })
      }
    );
  })
);
const FilterField = Object.assign(InternalFilterField, {
  Suggestions: import_Suggestions.Suggestions,
  Suggestion: import_Suggestion.Suggestion,
  SuggestionGroup: import_SuggestionGroup.SuggestionGroup,
  SuggestionGroupLabel: import_SuggestionGroupLabel.SuggestionGroupLabel,
  SuggestionDetails: import_SuggestionDetails.SuggestionDetails
});
