// @dynatrace/strato-components/filters/filter-field/FilterField.sty.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var FilterField_css_exports = {};
__export(FilterField_css_exports, {
  ariaDisabled: () => ariaDisabled,
  container: () => container,
  disabled: () => disabled,
  error: () => error,
  errorText: () => errorText,
  errorVars: () => errorVars,
  escapedCharacter: () => escapedCharacter,
  existsOperator: () => existsOperator,
  filterIcon: () => filterIcon,
  filterKey: () => filterKey,
  filterStatement: () => filterStatement,
  filterStatementBase: () => filterStatementBase,
  filterStatementDeleteButton: () => filterStatementDeleteButton,
  filterStatementDeleteIcon: () => filterStatementDeleteIcon,
  filterStatementDeleteWidget: () => filterStatementDeleteWidget,
  filterValue: () => filterValue,
  incompleteFilterStatement: () => incompleteFilterStatement,
  intermediateFilterStatement: () => intermediateFilterStatement,
  operator: () => operator,
  pastedContent: () => pastedContent,
  postSpotError: () => postSpotError,
  preSpotError: () => preSpotError,
  readOnly: () => readOnly,
  spotError: () => spotError,
  suggestionHintContainer: () => suggestionHintContainer,
  suggestionsOverlayAnnouncement: () => suggestionsOverlayAnnouncement,
  variable: () => variable,
  wrapper: () => wrapper
});
module.exports = __toCommonJS(FilterField_css_exports);
var import_AppRoot_css_ts_vanilla = require("../../core/components/app-root/AppRoot.css");
var import_FilterField_css_ts_vanilla = require("./FilterField.css");
var import_createRuntimeFn = require("@vanilla-extract/recipes/createRuntimeFn");
var ariaDisabled = "_96tj13o-3-6-0";
var container = "_96tj13c-3-6-0";
var disabled = "_96tj13p-3-6-0 _96tj13n-3-6-0 _96tj13o-3-6-0";
var error = "_96tj1311-3-6-0";
var errorText = "_96tj13e-3-6-0";
var errorVars = "_96tj13d-3-6-0";
var escapedCharacter = "_96tj1310-3-6-0";
var existsOperator = "_96tj13z-3-6-0";
var filterIcon = (0, import_createRuntimeFn.createRuntimeFn)({ defaultClassName: "_96tj13f-3-6-0", variantClassNames: { hasError: { true: "_96tj13g-3-6-0" }, disabled: { true: "_96tj13h-3-6-0" } }, defaultVariants: { hasError: false, disabled: false }, compoundVariants: [] });
var filterKey = "_96tj13v-3-6-0";
var filterStatement = "_96tj13j-3-6-0 _96tj13i-3-6-0";
var filterStatementBase = "_96tj13i-3-6-0";
var filterStatementDeleteButton = (0, import_createRuntimeFn.createRuntimeFn)({ defaultClassName: "_96tj13r-3-6-0", variantClassNames: { incomplete: { true: "_96tj13s-3-6-0" } }, defaultVariants: {}, compoundVariants: [] });
var filterStatementDeleteIcon = "_96tj13t-3-6-0";
var filterStatementDeleteWidget = "_96tj13q-3-6-0";
var filterValue = "_96tj13w-3-6-0";
var incompleteFilterStatement = "_96tj13k-3-6-0 _96tj13i-3-6-0";
var intermediateFilterStatement = "_96tj13l-3-6-0";
var operator = "_96tj13y-3-6-0";
var pastedContent = "_96tj13m-3-6-0";
var postSpotError = "_96tj1314-3-6-0";
var preSpotError = "_96tj1313-3-6-0";
var readOnly = "_96tj13n-3-6-0";
var spotError = "_96tj1312-3-6-0";
var suggestionHintContainer = "_96tj13u-3-6-0";
var suggestionsOverlayAnnouncement = "_96tj1315-3-6-0";
var variable = "_96tj13x-3-6-0";
var wrapper = "_96tj13b-3-6-0";
