// @dynatrace/strato-components/notifications/toast/Toast.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Toast_exports = {};
__export(Toast_exports, {
  dismissAllToasts: () => dismissAllToasts,
  dismissToast: () => dismissToast,
  showToast: () => showToast
});
module.exports = __toCommonJS(Toast_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_toastify = require("react-toastify");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_Toast_css = require("./Toast.sty.js");
var import_OverlayStacking = require("../../core/contexts/OverlayStacking.js");
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_Container = require("../../layouts/container/Container.js");
var import_Flex = require("../../layouts/flex/Flex.js");
var import_textStyle_css = require("../../styles/textStyle.sty.js");
var import_Text = require("../../typography/text/Text.js");
const COMPONENT_NAME = "Toast";
const Toast = (props) => {
  const {
    children,
    type,
    title,
    message,
    closeToast,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const ariaLabelingProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(restProps);
  const Icon = (0, import_react.useMemo)(() => {
    switch (type) {
      case "warning":
        return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.WarningIcon, {});
      case "success":
        return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.SuccessIcon, {});
      case "critical":
        return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.CriticalIcon, {});
      case "info":
        return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.InformationIcon, {});
    }
  }, [type]);
  const closeToastOnEsc = (e) => {
    if (e.key === "Escape") {
      e.preventDefault();
      closeToast();
    }
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      onKeyDown: closeToastOnEsc,
      className: (0, import_clsx.default)(consumerClassName, import_Toast_css.toastCSS),
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...ariaLabelingProps,
      ...behavioralTrackingProps,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_OverlayStacking.OverlayStackingContext.Provider, { value: 100, children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_Container.Container,
          {
            variant: "accent",
            color: type === "info" ? "primary" : type,
            className: import_Toast_css.toastIconCSS,
            children: Icon
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
          import_Flex.Flex,
          {
            className: (0, import_clsx.default)(
              import_Toast_css.toastFlexContainerCSS,
              (0, import_textStyle_css.textStyle)({ textStyle: "base" })
            ),
            paddingLeft: 16,
            paddingTop: 12,
            paddingBottom: 12,
            paddingRight: 8,
            flexDirection: "column",
            alignItems: "flex-start",
            flexGrow: 1,
            gap: 6,
            children: [
              title && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { className: import_Toast_css.toastTextCSS, textStyle: "base-emphasized", as: "p", children: title }),
              message && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { className: import_Toast_css.toastTextCSS, textStyle: "base", as: "p", children: message }),
              children
            ]
          }
        )
      ] })
    }
  );
};
function showToast({
  id,
  type = "info",
  role,
  lifespan,
  title,
  message,
  position = "bottom-left",
  actions,
  className: consumerClassName,
  style: consumerStyle,
  "aria-label": ariaLabel,
  "aria-labelledby": ariaLabelledBy,
  "aria-describedby": ariaDescribedBy,
  "aria-details": ariaDetails,
  "data-testid": dataTestId,
  "data-dtrum-mask": dataDtrumMask,
  "data-dtrum-allow": dataDtrumAllow,
  ...remainingProps
}) {
  const defaultRole = type === "critical" ? "alert" : "status";
  const defaultAutoClose = type === "warning" || type === "critical" ? false : 8e3;
  const autoClose = lifespan === "infinite" ? false : lifespan;
  return (0, import_react_toastify.toast)(
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      Toast,
      {
        type,
        title,
        message,
        closeToast: () => void 0,
        className: consumerClassName,
        style: consumerStyle,
        "aria-label": ariaLabel,
        "aria-labelledby": ariaLabelledBy,
        "aria-describedby": ariaDescribedBy,
        "aria-details": ariaDetails,
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        ...remainingProps,
        "data-dt-component": COMPONENT_NAME,
        children: actions
      }
    ),
    {
      toastId: id,
      autoClose: autoClose ?? defaultAutoClose,
      role: role ?? defaultRole,
      position
    }
  );
}
function dismissToast(id) {
  import_react_toastify.toast.dismiss(id);
}
function dismissAllToasts() {
  import_react_toastify.toast.dismiss();
}
