// @dynatrace/strato-components/notifications/toast/ToastContainer.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ToastContainer_exports = {};
__export(ToastContainer_exports, {
  ToastContainer: () => ToastContainer
});
module.exports = __toCommonJS(ToastContainer_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_dom = require("react-dom");
var import_react_intl = require("react-intl");
var import_react_toastify = require("react-toastify");
var import_with_selector = require("use-sync-external-store/shim/with-selector");
var import_strato_icons = require("@dynatrace/strato-icons");
var styles = __toESM(require("./ToastContainer.sty.js"));
var import_Button = require("../../buttons/button/Button.js");
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_useSafeForwardProps = require("../../core/hooks/useSafeForwardProps.js");
var import_store = require("../../core/store/store.js");
var import_Flex = require("../../layouts/flex/Flex.js");
var import_useKeyboardFocusHandling = require("../../overlays/hooks/useKeyboardFocusHandling.js");
const COMPONENT_NAME = "ToastContainer";
let unique = 0;
const toastContainerStore = (0, import_store.createStore)({
  currentContainerId: null
});
const ToastContainer = (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const {
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...remainingProps
    } = props;
    const closeButtonRef = (0, import_react.useRef)(null);
    const containerRef = (0, import_react.useRef)(null);
    const ariaLabelingProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(remainingProps);
    const forwardProps = (0, import_useSafeForwardProps.useSafeForwardProps)(remainingProps, ariaLabelingProps);
    const { "aria-label": ariaLabel, ...remainingAriaLabelingProps } = ariaLabelingProps;
    const [handleFocusOnKeydown, closedOnKeydown] = (0, import_useKeyboardFocusHandling.useKeyboardFocusHandling)(
      closeButtonRef,
      containerRef
    );
    const rootRef = (0, import_useMergeRefs.useMergeRefs)([forwardedRef, containerRef]);
    const CloseButton = ({
      closeToast
    }) => {
      const intl = (0, import_react_intl.useIntl)();
      const closeButtonLabel = intl.formatMessage({
        defaultMessage: "Close notification",
        id: "7IzOa/lx4Dc/qLGo",
        description: "Aria-label for the close notification button."
      });
      const closeToastOnEsc = (event) => {
        if (event.key === "Escape") {
          event.preventDefault();
          closedOnKeydown.current = true;
          closeToast(event);
        }
      };
      const detectCloseOnKeyDown = (0, import_react.useCallback)(() => {
        closedOnKeydown.current = true;
      }, []);
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_Flex.Flex,
        {
          flexDirection: "column",
          alignItems: "flex-start",
          paddingTop: 6,
          paddingRight: 6,
          onKeyDown: closeToastOnEsc,
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_Button.Button,
            {
              onKeyDown: detectCloseOnKeyDown,
              onClick: closeToast,
              "aria-label": closeButtonLabel,
              ref: closeButtonRef,
              variant: "default",
              onPointerDown: (e) => e.stopPropagation(),
              "data-dt-name": "CloseX",
              children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.XmarkIcon, {}) })
            }
          )
        }
      );
    };
    const containerId = (0, import_react.useRef)(++unique);
    const activeId = (0, import_with_selector.useSyncExternalStoreWithSelector)(
      toastContainerStore.subscribe,
      () => toastContainerStore.getState(),
      null,
      (state) => state.currentContainerId
    );
    (0, import_react.useEffect)(() => {
      const id = containerId.current;
      if (toastContainerStore.getState().currentContainerId === null) {
        toastContainerStore.setState({ currentContainerId: id });
      }
    }, [activeId]);
    (0, import_react.useEffect)(() => {
      const id = containerId.current;
      let pendingObserver = null;
      const unsubscribeOnChange = import_react_toastify.toast.onChange((payload) => {
        if (payload.status === "removed") {
          const container = containerRef.current;
          if (!container) {
            return;
          }
          pendingObserver?.disconnect();
          pendingObserver = new MutationObserver((mutations) => {
            for (const mutation of mutations) {
              if (mutation.removedNodes.length > 0) {
                pendingObserver?.disconnect();
                pendingObserver = null;
                handleFocusOnKeydown();
                return;
              }
            }
          });
          pendingObserver.observe(container, {
            childList: true,
            subtree: true
          });
        }
      });
      return function cleanup() {
        unsubscribeOnChange();
        pendingObserver?.disconnect();
        if (toastContainerStore.getState().currentContainerId === id) {
          toastContainerStore.setState({ currentContainerId: null });
        }
      };
    }, []);
    if (activeId !== containerId.current) {
      return null;
    }
    return (0, import_react_dom.createPortal)(
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        "div",
        {
          ref: rootRef,
          className: (0, import_clsx.default)(consumerClassName, styles.root),
          style: consumerStyle,
          "data-toast-container": true,
          "data-testid": dataTestId,
          "data-dtrum-mask": dataDtrumMask,
          "data-dtrum-allow": dataDtrumAllow,
          ...remainingAriaLabelingProps,
          ...forwardProps,
          "data-dt-component": COMPONENT_NAME,
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_react_toastify.ToastContainer,
            {
              hideProgressBar: true,
              draggable: false,
              closeButton: CloseButton,
              limit: 5,
              pauseOnHover: true,
              closeOnClick: false,
              newestOnTop: false,
              ...ariaLabel && { "aria-label": ariaLabel },
              transition: (0, import_react_toastify.cssTransition)({
                collapseDuration: 200,
                enter: "strato__toast__enter",
                exit: "strato__toast__exit"
              })
            }
          )
        }
      ),
      document.body
    );
  }
);
ToastContainer.displayName = COMPONENT_NAME;
