// @dynatrace-sdk/app-environment/testing/cjs/index.js
/**
 * @license
 * Copyright 2023-2025 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/platform/app-environment/testing/index.ts
var testing_exports = {};
__export(testing_exports, {
  clearAppEnvironmentMocks: () => clearAppEnvironmentMocks,
  mockAppEnvironment: () => mockAppEnvironment,
  setupTestingEnv: () => setupTestingEnv
});
module.exports = __toCommonJS(testing_exports);

// packages/platform/app-environment/testing/test-utils.ts
var currentSetup;
var isInitialised = false;
var wrapper = {
  getAppId: () => currentSetup.getAppId(),
  getAppName: () => currentSetup.getAppName(),
  getAppVersion: () => currentSetup.getAppVersion(),
  getEnvironmentId: () => currentSetup.getEnvironmentId(),
  getEnvironmentUrl: () => currentSetup.getEnvironmentUrl(),
  getCurrentUserDetails: () => currentSetup.getCurrentUserDetails()
};
var setupTestingEnv = () => {
  if (typeof expect === "undefined") {
    isInitialised = true;
  }
  mockAppEnvironment();
};
var mockAppEnvironment = (setup = {}) => {
  if (!isInitialised) {
    throw new Error(`\`@dynatrace-sdk/app-environment/testing\` has not been initialised!
Add in your \`jest.config.js\` file entry with \`@dynatrace-sdk/app-environment/testing\` setup:

\`setupFiles: ['@dynatrace-sdk/app-environment/testing']\``);
  }
  const globalWithDtRuntimeAppEnvironment = globalThis;
  if (globalWithDtRuntimeAppEnvironment.dtRuntime === void 0) {
    globalWithDtRuntimeAppEnvironment.dtRuntime = {};
  }
  if (globalWithDtRuntimeAppEnvironment.dtRuntime.appEnvironment === void 0) {
    globalWithDtRuntimeAppEnvironment.dtRuntime.appEnvironment = wrapper;
  }
  const legacyDtSdk = globalThis;
  legacyDtSdk.dtSdk = legacyDtSdk.dtSdk || {};
  legacyDtSdk.dtSdk.getAppId = wrapper.getAppId;
  legacyDtSdk.dtSdk.getAppName = wrapper.getAppName;
  legacyDtSdk.dtSdk.getAppVersion = wrapper.getAppVersion;
  legacyDtSdk.dtSdk.getEnvironmentId = wrapper.getEnvironmentId;
  legacyDtSdk.dtSdk.getEnvironmentUrl = wrapper.getEnvironmentUrl;
  currentSetup = getSetupWithDefaults(setup);
};
var clearAppEnvironmentMocks = () => {
  mockAppEnvironment();
};
var getSetupWithDefaults = (setup) => {
  return {
    getAppId: setup.getAppId || (() => "mocked-app-id"),
    getAppName: setup.getAppName || (() => "mocked-app-name"),
    getAppVersion: setup.getAppVersion || (() => "mocked-app-version"),
    getEnvironmentId: setup.getEnvironmentId || (() => "mocked-environment-id"),
    getEnvironmentUrl: setup.getEnvironmentUrl || (() => "mocked-environment-url"),
    getCurrentUserDetails: setup.getCurrentUserDetails || (() => ({
      id: "mocked-user-id",
      name: "mocked-user-name",
      email: "mocked-user-email"
    }))
  };
};

// packages/platform/app-environment/testing/index.ts
setupTestingEnv();
