// @dynatrace/strato-components/esm/filters/filter/types/filter-state.types.js
const PinnedStateValues = [
  /** The item is not displayed in the filter bar, but is available in the add-filter select and can change to a "pinned-optional" item at any time. */
  "optional",
  /** The item is currently displayed in the filter bar, but is available in the add-filter select and can change to an "optional" item at any time. */
  "pinned-optional",
  /** The item is always in the filter bar and not controllable with the add-filter select. Changing from "pinned" to any other state is also not recommended since that might break UX patterns.*/
  "pinned"
];
var FilterBarActions = /* @__PURE__ */ ((FilterBarActions2) => {
  FilterBarActions2["INITIALIZE"] = "initialize";
  FilterBarActions2["CHANGE"] = "change";
  FilterBarActions2["REMOVE"] = "remove";
  FilterBarActions2["PIN"] = "pin";
  FilterBarActions2["UNPIN"] = "unpin";
  FilterBarActions2["CHANGEPINNEDSTATE"] = "changepinnedstate";
  return FilterBarActions2;
})(FilterBarActions || {});
export {
  FilterBarActions,
  PinnedStateValues
};
