// @dynatrace/strato-components/esm/charts/core/components/single-value-renderer/utils/build-value-content-grid-rows.js
const fractions = {
  label: 3,
  value: 6,
  trend: 3,
  sparkline: 8
};
const sumFractions = (configWithFractions) => configWithFractions.reduce((acc, cur) => acc + cur, 0);
const componentsHeightMapToCSSGridTemplateString = (fractionsMap) => Array.from(fractionsMap).map(([key, value]) => `"${key}" ${value}`).join("\r\n");
const getInnerGridValue = (innerContentMap, localFractionSum) => {
  const innerGridPercentValues = /* @__PURE__ */ new Map();
  let allAreasHeight = 0;
  innerContentMap.forEach((value, key) => {
    const areaHeight = Math.round(value / localFractionSum * 100);
    allAreasHeight += areaHeight;
    const isRoundedOnePercentLeft = 100 - allAreasHeight === 1;
    innerGridPercentValues.set(
      key,
      `${isRoundedOnePercentLeft ? areaHeight + 1 : areaHeight}cqh`
    );
  });
  return componentsHeightMapToCSSGridTemplateString(innerGridPercentValues);
};
const getOuterGridValue = (outerContentMap, fractionsSum) => {
  const outerGridPercentValues = /* @__PURE__ */ new Map();
  outerContentMap.forEach(
    (value, key) => outerGridPercentValues.set(
      key,
      `${Math.round(value / fractionsSum * 100)}cqh`
    )
  );
  const valueHeight = 100 - parseInt(outerGridPercentValues.get("label") ?? "0");
  outerGridPercentValues.set("content", `${Math.round(valueHeight)}cqh`);
  return componentsHeightMapToCSSGridTemplateString(outerGridPercentValues);
};
function buildValueContentGridRows(hasLabel, hasValue, hasTrend, hasSparkline) {
  const innerComponents = /* @__PURE__ */ new Map();
  const outerComponents = /* @__PURE__ */ new Map();
  hasValue && innerComponents.set("value", fractions.value);
  hasTrend && innerComponents.set("trend", fractions.trend);
  hasSparkline && innerComponents.set("sparkline", fractions.sparkline);
  hasLabel && outerComponents.set("label", fractions.label);
  const innerComponentsFractionsArray = [...innerComponents.values()];
  const innerFractionSum = sumFractions(innerComponentsFractionsArray);
  const outerFractionSum = sumFractions([
    ...innerComponentsFractionsArray,
    ...outerComponents.values()
  ]);
  return {
    innerGrid: getInnerGridValue(innerComponents, innerFractionSum),
    outerGrid: getOuterGridValue(outerComponents, outerFractionSum)
  };
}
export {
  buildValueContentGridRows,
  componentsHeightMapToCSSGridTemplateString,
  getInnerGridValue,
  sumFractions
};
