// @dynatrace/strato-components/esm/charts/core/components/overlay-tooltip/utils/build-content.js
import { jsx } from "react/jsx-runtime";
import { clone } from "lodash-es";
import { Children, Fragment } from "react";
import { isSlot } from "../../../../../core/utils/is-slot.js";
import { Footer } from "../components/Footer.js";
import { Action } from "../components/global-actions/Action.js";
import { ActionsPlaceholder } from "../components/global-actions/ActionsPlaceholder.js";
import { Header } from "../components/header/Header.js";
import { Series } from "../components/series/Series.js";
import { SeriesHeader } from "../components/series/SeriesHeader.js";
import { SeriesItem } from "../components/series/SeriesItem.js";
import { DEFAULT_TRUNCATION_MODE } from "../constants.js";
const VALID_TRUNCATION_MODES = ["start", "middle", "end"];
const isValidTruncationMode = (value) => {
  return typeof value === "string" && VALID_TRUNCATION_MODES.includes(value);
};
const getTruncationMode = (userTruncationMode, defaultConfigTruncationMode) => {
  if (isValidTruncationMode(userTruncationMode)) {
    return userTruncationMode;
  }
  if (isValidTruncationMode(defaultConfigTruncationMode)) {
    return defaultConfigTruncationMode;
  }
  return DEFAULT_TRUNCATION_MODE;
};
function buildContent(children, defaultConfig, args) {
  const slots = {
    Header: null,
    Footer: null,
    Series: null,
    SeriesHeader: null,
    SeriesItem: null,
    ActionsPlaceholder: null,
    GlobalActions: []
  };
  let unpinnedPlaceholderText;
  let defaultActionsConfig = defaultConfig?.defaultActionsConfig;
  if (!args.customTooltip) {
    Children.forEach(children, (child) => {
      if (isSlot(child, Header)) {
        slots.Header = child;
      } else if (isSlot(child, Footer)) {
        slots.Footer = child;
      } else if (isSlot(child, SeriesHeader)) {
        slots.SeriesHeader = child;
        defaultActionsConfig = clone(
          defaultConfig?.defaultActionsConfig
        );
        const defaultNameAction = defaultActionsConfig?.get("name");
        if (child.props.seriesLabel && defaultNameAction) {
          defaultActionsConfig?.set("name", {
            ...defaultNameAction,
            customLabel: child.props.seriesLabel
          });
        }
        const defaultValueAction = defaultActionsConfig?.get("value");
        if (child.props.valueLabel && defaultValueAction) {
          defaultActionsConfig?.set("value", {
            ...defaultValueAction,
            customLabel: child.props.valueLabel
          });
        }
      } else if (isSlot(child, Series)) {
        slots.Series = child;
      } else if (isSlot(child, SeriesItem)) {
        slots.SeriesItem = child;
      } else if (isSlot(child, ActionsPlaceholder)) {
        unpinnedPlaceholderText = child.props.text;
      } else if (isSlot(child, Action)) {
        slots.GlobalActions.push(child);
      }
    });
  }
  return {
    content: [
      /* @__PURE__ */ jsx(Fragment, { children: slots.Header }, "header"),
      slots.SeriesItem ? /* @__PURE__ */ jsx(Fragment, { children: slots.SeriesHeader || /* @__PURE__ */ jsx(SeriesHeader, {}) }, "seriesHeader") : null,
      /* @__PURE__ */ jsx(Fragment, { children: slots.Series || slots.SeriesItem }, "series"),
      slots.Footer ?? /* @__PURE__ */ jsx(
        Footer,
        {
          actions: slots.GlobalActions,
          unpinnedPlaceholderText
        },
        "footer"
      )
    ],
    config: {
      seriesHeader: slots.SeriesHeader || /* @__PURE__ */ jsx(SeriesHeader, {}),
      displayHeaderDivider: !!(slots.Header && (slots.Series || slots.SeriesItem)),
      symbolPosition: args.userSymbolPosition || defaultConfig?.symbolPosition || "left",
      truncationMode: getTruncationMode(
        args.userTruncationMode,
        defaultConfig?.truncationMode
      ),
      defaultActionsConfig,
      variant: args.customTooltip || defaultConfig?.sharedVariant && slots.Series?.props.variant === "shared" ? "shared" : "single"
    }
  };
}
export {
  buildContent
};
