// @dynatrace/strato-geo/esm/map/components/DotLayer/utils/parse-dot-data-to-geo-json.js
import { _getCanvasColor as getCanvasColor } from "@dynatrace/strato-components/charts";
import { defineBearing } from "./define-bearing.js";
import { resolveIconBackground } from "./resolve-icon-background.js";
import { calculateHoveredColor } from "../../../utils/calculate-hovered-color.js";
import { defineColor } from "../../../utils/define-color.js";
const parseDotDataToGeoJSON = (data, color, bearing, backgroundColor) => {
  return {
    type: "FeatureCollection",
    features: data.map((feature) => {
      const { longitude, latitude, ...properties } = feature;
      const canvasColor = getCanvasColor(defineColor(color, feature));
      const hoveredColor = calculateHoveredColor(canvasColor);
      const { parsedBackgroundColor, parsedHoveredBackgroundColor } = resolveIconBackground(feature, backgroundColor);
      return {
        type: "Feature",
        properties: {
          data: feature,
          ...properties,
          __color: canvasColor,
          __hoveredColor: hoveredColor,
          __backgroundColor: parsedBackgroundColor,
          __hoveredBackgroundColor: parsedHoveredBackgroundColor,
          __bearing: defineBearing(bearing, feature)
        },
        geometry: { type: "Point", coordinates: [longitude, latitude] }
      };
    })
  };
};
export {
  parseDotDataToGeoJSON
};
