// @dynatrace/strato-components/esm/charts/tree-map/TreeMap.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import { isEmpty } from "lodash-es";
import {
  forwardRef,
  useImperativeHandle,
  useRef
} from "react";
import { TreeMapChartLayout } from "./components/TreeMapChartLayout.js";
import { DEFAULT_HEIGHT, DEFAULT_WIDTH } from "./constants.js";
import { useTreeMapChartConfig } from "./hooks/useTreeMapChartConfig.js";
import { TreeMapChartProviders } from "./providers/TreeMapChartProviders.js";
import { TreeMapDownloadCSVSlot } from "./slots/DownloadCSV.js";
import { Toolbar } from "./slots/Toolbar.js";
import { Tooltip } from "./slots/Tooltip.js";
import { Legend as TreeMapChartLegend } from "./slots/TreeMapChartLegend.js";
import { buildAndDownloadCsv } from "./utils/build-and-download-csv.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { Intent } from "../../core/slots/Intent/Intent.js";
import { defaultChartSize } from "../../core/styles/defaultChartValues.sty.js";
import { useAutoLegendRefresh } from "../core/components/chart-layout/hooks/useAutoLegendRefresh.js";
import { getChartStatesTemplates } from "../core/components/chart-states/utils/get-chart-states-template.js";
import { EmptyStateSlot } from "../core/slots/chart-states/EmptyStateSlot.js";
import { ErrorStateSlot } from "../core/slots/chart-states/ErrorStateSlot.js";
import { ColorRule } from "../core/slots/ColorRule.js";
import { ChartType } from "../core/types/chart-types.js";
const COMPONENT_NAME = "TreeMap";
const TreeMapRoot = forwardRef((props, forwardedRef) => {
  const {
    data,
    children,
    seriesActions,
    width = DEFAULT_WIDTH,
    height = DEFAULT_HEIGHT,
    className,
    style,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "data-testid": dataTestId = "treemap-chart-root",
    unit
  } = props;
  const [behavioralTrackingProps] = useBehavioralTrackingProps(props);
  const config = useTreeMapChartConfig(props);
  const { loading } = config;
  const emptyData = isEmpty(data) || data.tree.nodes === void 0 || data.tree.nodes.length === 0;
  const { emptyState, errorState } = getChartStatesTemplates(children);
  const treeMapRef = useRef(null);
  const chartLayoutRef = useAutoLegendRefresh(data);
  useImperativeHandle(forwardedRef, () => {
    let exposedRef = {};
    if (treeMapRef.current) {
      exposedRef = {
        element: treeMapRef.current,
        getConfig: () => JSON.stringify({ type: ChartType.Treemap, ...config }),
        downloadData: () => buildAndDownloadCsv(data.tree)
      };
    }
    return exposedRef;
  });
  return /* @__PURE__ */ jsx(
    TreeMapChartProviders,
    {
      data,
      seriesActions,
      unit,
      ...config,
      children: /* @__PURE__ */ jsx(
        TreeMapChartLayout,
        {
          ...behavioralTrackingProps,
          chartLayoutRef,
          ref: treeMapRef,
          width,
          height,
          emptyState,
          errorState,
          isEmpty: emptyData,
          loading,
          className: clsx(defaultChartSize, className),
          style,
          "data-dtrum-mask": dataDtrumMask,
          "data-dtrum-allow": dataDtrumAllow,
          "data-testid": dataTestId,
          "data-dt-component": COMPONENT_NAME
        }
      )
    }
  );
});
const TreeMap = Object.assign(TreeMapRoot, {
  Legend: TreeMapChartLegend,
  Tooltip,
  Intent,
  Toolbar,
  DownloadCSV: TreeMapDownloadCSVSlot,
  EmptyState: EmptyStateSlot,
  ErrorState: ErrorStateSlot,
  ColorRule
});
export {
  TreeMap
};
