// @dynatrace/strato-components/esm/editors/base-code-editor/configuration/language-extensions.js
import { autocompletion } from "@codemirror/autocomplete";
import { StreamLanguage } from "@codemirror/language";
import {
  autocompletionExtension
} from "../auto-completion/autocompletion.extension.js";
import { listAutocompletionExtension } from "../auto-completion/list/list-autocompletion-extension.js";
const emptyLanguage = StreamLanguage.define({
  token: (stream) => {
    stream.skipToEnd();
    return null;
  }
});
function getLanguageExtensions(language, providers, intlProps) {
  const configLanguage = language ?? emptyLanguage;
  return [
    configLanguage,
    ...getAutocompleteExtensions(providers || [], intlProps).map(
      (autocomplete) => configLanguage.data.of({ autocomplete })
    )
  ];
}
function getAutocompleteExtensions(providers, intlProps) {
  return providers.map(
    (p) => typeof p === "function" ? autocompletionExtension(p, intlProps) : listAutocompletionExtension(p, intlProps)
  );
}
function getLanguageConfiguration(languageConfig, intlProps) {
  let providers;
  if (languageConfig.autocompleteOptions) {
    providers = languageConfig.autocompleteOptions.providers;
  }
  return [
    getLanguageExtensions(languageConfig.language, providers, intlProps),
    autocompletion({
      activateOnTyping: languageConfig.autocompleteOptions?.activateOnTyping,
      optionClass: languageConfig.autocompleteOptions?.optionClass
    }),
    ...languageConfig.extendWith ?? []
  ];
}
export {
  getAutocompleteExtensions,
  getLanguageConfiguration,
  getLanguageExtensions
};
