// @dynatrace/strato-components/charts/gauge/providers/GaugeChartProviders.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var GaugeChartProviders_exports = {};
__export(GaugeChartProviders_exports, {
  GaugeChartProviders: () => GaugeChartProviders
});
module.exports = __toCommonJS(GaugeChartProviders_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_OverlayContext = require("../../core/components/overlay-tooltip/contexts/OverlayContext.js");
var import_TooltipContext = require("../../core/components/overlay-tooltip/contexts/TooltipContext.js");
var import_ChartSeriesActionContext = require("../../core/components/series-action/ChartSeriesActionContext.js");
var import_GaugeChartColorContext = require("../contexts/GaugeChartColorContext.js");
var import_GaugeChartFormatterContext = require("../contexts/GaugeChartFormatterContext.js");
var import_GaugeChartLabelsContext = require("../contexts/GaugeChartLabelsContext.js");
var import_GaugeChartLoadingContext = require("../contexts/GaugeChartLoadingContext.js");
var import_GaugeChartRangeMaxContext = require("../contexts/GaugeChartRangeMaxContext.js");
var import_GaugeChartRangeMinContext = require("../contexts/GaugeChartRangeMinContext.js");
var import_GaugeChartSizeContext = require("../contexts/GaugeChartSizeContext.js");
var import_GaugeChartThresholdContext = require("../contexts/GaugeChartThresholdContext.js");
var import_GaugeChartTooltipContext = require("../contexts/GaugeChartTooltipContext.js");
var import_get_value_arc_color = require("../utils/get-value-arc-color.js");
const emptyThresholds = [];
const GaugeChartProviders = ({
  config,
  width,
  height,
  value,
  valueAccessor,
  explicitColor,
  seriesActions,
  children,
  loading
}) => {
  const {
    formatter,
    max,
    min,
    labelSlot,
    thresholdsIndicators,
    colorRules,
    unit
  } = config;
  const calculatedColor = (0, import_get_value_arc_color.getValueArcColor)(
    value,
    colorRules,
    explicitColor,
    thresholdsIndicators,
    valueAccessor
  );
  const seriesActionValue = (0, import_react.useMemo)(
    () => ({
      chartSeriesActions: seriesActions
    }),
    [seriesActions]
  );
  const chartSizeValue = (0, import_react.useMemo)(() => ({ width, height }), [width, height]);
  const formatterContextValue = (0, import_react.useMemo)(
    () => ({
      formatter,
      unit
    }),
    [formatter, unit]
  );
  const tooltipConfigValue = (0, import_react.useMemo)(
    () => ({
      name: config.tooltip?.name,
      template: config.tooltip?.children,
      hidden: config.tooltip?.hidden ?? false
    }),
    [config.tooltip]
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayContext.OverlayProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TooltipContext.TooltipProvider, { persistedStore: true, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ChartSeriesActionContext.ChartSeriesActionContext.Provider, { value: seriesActionValue, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_GaugeChartSizeContext.GaugeChartSizeContext.Provider, { value: chartSizeValue, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_GaugeChartFormatterContext.GaugeChartFormatterContext.Provider, { value: formatterContextValue, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_GaugeChartThresholdContext.GaugeChartThresholdContext.Provider,
    {
      value: thresholdsIndicators ?? emptyThresholds,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_GaugeChartColorContext.GaugeChartColorContext.Provider, { value: calculatedColor, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_GaugeChartLabelsContext.GaugeChartLabelsContext.Provider,
        {
          value: {
            label: labelSlot
          },
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_GaugeChartRangeMinContext.GaugeChartRangeMinContext.Provider, { value: min, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_GaugeChartRangeMaxContext.GaugeChartRangeMaxContext.Provider, { value: max, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_GaugeChartLoadingContext.GaugeChartLoadingContext.Provider, { value: loading, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_GaugeChartTooltipContext.GaugeChartTooltipContext.Provider,
            {
              value: tooltipConfigValue,
              children
            }
          ) }) }) })
        }
      ) })
    }
  ) }) }) }) }) });
};
GaugeChartProviders["displayName"] = "GaugeChartProviders";
