// @dynatrace/strato-components/charts/core/components/annotations/components/canvas/AnnotationsCanvasLayout.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var AnnotationsCanvasLayout_exports = {};
__export(AnnotationsCanvasLayout_exports, {
  AnnotationsCanvasLayout: () => AnnotationsCanvasLayout
});
module.exports = __toCommonJS(AnnotationsCanvasLayout_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_AnnotationsCanvasLayout_css = require("./AnnotationsCanvasLayout.sty.js");
var import_display_config_context = require("./context/display-config.context.js");
var import_AnnotationsChartDisplay = require("./displays/AnnotationsChartDisplay.js");
var import_Canvas = require("../../../canvas/Canvas.js");
var import_CanvasContext = require("../../../canvas/CanvasContext.js");
var import_useAnnotationsIconImages = require("../../hooks/useAnnotationsIconImages.js");
const AnnotationsCanvasLayout = ({
  height,
  width
}) => {
  const icons = (0, import_useAnnotationsIconImages.useAnnotationsIconImages)();
  const annotationsArea = { width, height, icons };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_display_config_context.DisplayConfigContext.Provider, { value: { annotationsArea }, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      className: import_AnnotationsCanvasLayout_css.canvasContainer,
      style: {
        width,
        height,
        overflowY: "hidden"
      },
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_CanvasContext.CanvasContext.Provider,
        {
          value: {
            shapeRefs: /* @__PURE__ */ new Map()
          },
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Canvas.Canvas, { width, height, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_AnnotationsChartDisplay.AnnotationsChartDisplay, {}) })
        }
      )
    }
  ) });
};
AnnotationsCanvasLayout["displayName"] = "AnnotationsCanvasLayout";
