// @dynatrace/strato-components/forms/spin-buttons/spin-button-group/SpinButtonGroup.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SpinButtonGroup_exports = {};
__export(SpinButtonGroup_exports, {
  SpinButtonGroup: () => SpinButtonGroup
});
module.exports = __toCommonJS(SpinButtonGroup_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_useNavigateSpinButton = require("./hooks/useNavigateSpinButton.js");
var import_useValueOverflow = require("./hooks/useValueOverflow.js");
var import_SpinButtonGroupSeparator = require("./SpinButtonGroupSeparator.js");
var import_useBehavioralTrackingProps = require("../../../core/hooks/useBehavioralTrackingProps.js");
var import_NumberSpinButton = require("../number-spin-button/NumberSpinButton.js");
var import_SpinButton_css = require("../SpinButton.sty.js");
var import_TextSpinButton = require("../text-spin-button/TextSpinButton.js");
const InternalSpinButtonGroup = ({
  children,
  valueOverflow = false,
  onLeadingEdge,
  onTrailingEdge,
  onTrailingOverflow,
  className: consumerClassName,
  style: consumerStyle,
  "data-testid": dataTestId,
  "data-dtrum-mask": dataDtrumMask,
  "data-dtrum-allow": dataDtrumAllow,
  ...remainingProps
}) => {
  const [behavioralTrackingProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const { clonedChildren, spinButtonRefs, triggerRefs } = (0, import_useValueOverflow.useValueOverflow)(
    children,
    valueOverflow,
    onTrailingOverflow,
    onTrailingEdge
  );
  const { onKeyDownHandler } = (0, import_useNavigateSpinButton.useNavigateSpinButton)(
    spinButtonRefs,
    triggerRefs,
    onLeadingEdge,
    onTrailingEdge
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "span",
    {
      className: (0, import_clsx.default)(import_SpinButton_css.spinButtonGroupCSS, consumerClassName),
      "data-valuetype": "iso8601",
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...behavioralTrackingProps,
      onKeyDown: onKeyDownHandler,
      children: clonedChildren
    }
  );
};
const SpinButtonGroup = Object.assign(InternalSpinButtonGroup, {
  Number: import_NumberSpinButton.NumberSpinButton,
  Text: import_TextSpinButton.TextSpinButton,
  Separator: import_SpinButtonGroupSeparator.SpinButtonGroupSeparator
});
SpinButtonGroup.displayName = "_SpinButtonGroup";
