// @dynatrace/strato-components/esm/charts/xy-chart/components/annotations/XYChartAnnotationsArea.js
import { jsx } from "react/jsx-runtime";
import { isUndefined } from "lodash-es";
import { useBehavioralTrackingProps } from "../../../../core/hooks/useBehavioralTrackingProps.js";
import { AnnotationsNumericalDisplay } from "../../../core/components/annotations/components/displays/AnnotationsNumericalDisplay.js";
import { AnnotationsTimeDisplay } from "../../../core/components/annotations/components/displays/AnnotationsTimeDisplay.js";
import { AnnotationsScrollWrapper } from "../../../core/components/annotations/components/overlays/AnnotationsScrollWrapper.js";
import {
  DEFAULT_INTEGRATION_TEXT_OVERFLOW,
  SCROLLBAR_THRESHOLD
} from "../../../core/components/annotations/constants.js";
import { ChartGrid } from "../../../core/components/chart-grid/ChartGridRoot.js";
import { useGridPlotSize } from "../../../core/components/chart-grid/hooks/useGridPlotSize.js";
import { ANNOTATIONS_PADDING_BOTTOM } from "../../constants/defaults.js";
import { useXYChartAnnotationsConfig } from "../../hooks/useXYChartAnnotationsConfig.js";
import { useXYChartAnnotationsData } from "../../hooks/useXYChartAnnotationsData.js";
import { useXYChartData } from "../../hooks/useXYChartData.js";
import { useXYChartScales } from "../../hooks/useXYChartScales.js";
import { useSetState } from "../../store/selectors.js";
import { isTimeScale } from "../../types/guards.js";
import { getCurrentScale } from "../../utils/get-scale.js";
import { AREAS } from "../../utils/grid-templates.js";
const XYChartAnnotationsArea = () => {
  const scales = useXYChartScales();
  const annotationsConfig = useXYChartAnnotationsConfig();
  const annotationsData = useXYChartAnnotationsData();
  const setState = useSetState();
  const { width } = useGridPlotSize();
  const { axes } = useXYChartData();
  const [behaviorTrackingProps] = useBehavioralTrackingProps(
    annotationsConfig ?? {}
  );
  const handleUpdateIndicators = (event) => {
    setState((prev) => ({
      ...prev,
      annotations: {
        ...prev.annotations,
        ...event
      }
    }));
  };
  if (isUndefined(annotationsConfig) || isUndefined(annotationsData)) {
    return null;
  }
  const { xAxisId, showLabels = false, annotationsActions } = annotationsConfig;
  const axis = axes.find(({ id }) => id === xAxisId);
  if (axis === void 0) {
    return null;
  }
  const scale = getCurrentScale(scales, xAxisId);
  const trackNames = Array.from(annotationsData.keys());
  const serializedAnnotationsData = trackNames.map((name) => annotationsData.get(name)).filter((track) => !isUndefined(track));
  const DisplayComponent = isTimeScale({
    axisType: axis.type,
    axisScale: scale
  }) ? AnnotationsTimeDisplay : AnnotationsNumericalDisplay;
  return /* @__PURE__ */ jsx(
    ChartGrid.GridArea,
    {
      paddingBottom: ANNOTATIONS_PADDING_BOTTOM,
      gridArea: AREAS.annotations,
      children: ({ availableWidth, availableHeight }) => {
        const labelsOffset = availableWidth - width;
        const annotationsCommonConfig = {
          tracks: serializedAnnotationsData,
          showLabels,
          labelsOffset,
          width: availableWidth,
          markerGroupHover: handleUpdateIndicators,
          textOverflow: DEFAULT_INTEGRATION_TEXT_OVERFLOW,
          truncateMode: "middle",
          tooltipConfig: annotationsConfig.tooltip,
          ...behaviorTrackingProps
        };
        return /* @__PURE__ */ jsx(
          AnnotationsScrollWrapper,
          {
            height: availableHeight,
            width: availableWidth + SCROLLBAR_THRESHOLD,
            children: /* @__PURE__ */ jsx(
              DisplayComponent,
              {
                scale,
                annotationsActions: (annotation) => annotation && annotationsActions?.(annotation),
                ...annotationsCommonConfig
              }
            )
          }
        );
      }
    }
  );
};
export {
  XYChartAnnotationsArea
};
