// @dynatrace/strato-components/esm/charts/gantt-chart/hooks/use-set-tooltip-state.js
import { useCallback, useContext } from "react";
import { useStoreSelector } from "../../../core/store/selector.js";
import {
  GanttChartStateContext
} from "../context/GanttChartStateContext.js";
const useSetTooltipState = () => {
  const { setState } = useContext(GanttChartStateContext);
  const setTooltipStateContent = useCallback(
    (content) => {
      setState((prev) => {
        return {
          ...prev,
          tooltip: {
            ...prev.tooltip,
            content
          }
        };
      });
    },
    [setState]
  );
  const setTooltipStatePosition = useCallback(
    (position) => {
      setState((prev) => {
        return {
          ...prev,
          tooltip: {
            ...prev.tooltip,
            position
          }
        };
      });
    },
    [setState]
  );
  const setTooltipStateVisibility = useCallback(
    (visible) => {
      setState((prev) => {
        return {
          ...prev,
          tooltip: {
            ...prev.tooltip,
            visible
          }
        };
      });
    },
    [setState]
  );
  const setTooltipStatePinned = useCallback(
    (pinned) => {
      setState((prev) => {
        return {
          ...prev,
          tooltip: {
            ...prev.tooltip,
            pinned
          }
        };
      });
    },
    [setState]
  );
  return {
    setTooltipStateContent,
    setTooltipStatePosition,
    setTooltipStateVisibility,
    setTooltipStatePinned
  };
};
const useSelector = (selector) => useStoreSelector(GanttChartStateContext, selector);
const useTooltipContent = () => useSelector((state) => state.tooltip.content);
const useTooltipPinnedStatus = () => useSelector((state) => state.tooltip.pinned);
const useTooltipPosition = () => useSelector((state) => state.tooltip.position);
const useTooltipVisibility = () => useSelector((state) => state.tooltip.visible);
export {
  useSetTooltipState,
  useTooltipContent,
  useTooltipPinnedStatus,
  useTooltipPosition,
  useTooltipVisibility
};
