// @dynatrace/strato-components/esm/charts/gantt-chart/hooks/use-gantt-entry-handlers.js
import {
  useSetTooltipState,
  useTooltipPinnedStatus
} from "./use-set-tooltip-state.js";
function useGanttEntryHandlers(content) {
  const isTooltipPinned = useTooltipPinnedStatus();
  const {
    setTooltipStateContent,
    setTooltipStatePinned,
    setTooltipStateVisibility,
    setTooltipStatePosition
  } = useSetTooltipState();
  if (content === void 0) {
    return {
      onMouseEnterHandler: void 0,
      onMouseClickHandler: void 0,
      onMouseLeaveHandler: void 0
    };
  }
  const onMouseEnterHandler = () => {
    setTooltipStateVisibility(true);
    setTooltipStateContent({
      segment: content.segment,
      row: content.row,
      parent: content.parent
    });
  };
  const onMouseMoveHandler = (event) => {
    const { clientX, clientY } = event;
    const position = {
      x: clientX,
      y: clientY
    };
    setTooltipStatePosition(position);
  };
  const onMouseClickHandler = () => {
    setTooltipStatePinned(false);
    setTooltipStateContent({
      segment: content.segment,
      row: content.row,
      parent: content.parent
    });
    requestAnimationFrame(() => setTooltipStatePinned(true));
  };
  const onMouseLeaveHandler = () => {
    !isTooltipPinned && setTooltipStateVisibility(false);
    setTooltipStateContent(void 0);
  };
  return {
    onMouseEnterHandler,
    onMouseClickHandler,
    onMouseLeaveHandler,
    onMouseMoveHandler
  };
}
export {
  useGanttEntryHandlers
};
