// @dynatrace/strato-components/tables/DataTable/features/LineWrap/LineWrap.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var LineWrap_exports = {};
__export(LineWrap_exports, {
  DataTableGlobalLineWrapAction: () => DataTableGlobalLineWrapAction,
  DataTableGlobalLineWrapToolbarAction: () => DataTableGlobalLineWrapToolbarAction,
  LineWrap: () => LineWrap,
  useColumnLineWrap: () => useColumnLineWrap
});
module.exports = __toCommonJS(LineWrap_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_table = require("@tanstack/react-table");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_Button = require("../../../../buttons/button/Button.js");
var import_useControllableState = require("../../../../core/hooks/useControllableState.js");
var import_Tooltip = require("../../../../overlays/tooltip/Tooltip.js");
var import_useDataTableConfig = require("../../../hooks/useDataTableConfig.js");
var import_DataTableContext = require("../../contexts/DataTableContext.js");
var import_messages = require("../../messages.js");
function DataTableGlobalLineWrapAction() {
  return null;
}
const LineWrap = {
  createColumn(column, table) {
    column.toggleLineWrap = (value) => {
      if (!column.getCanLineWrap()) {
        return;
      }
      table.setColumnLineWrap((oldValue) => {
        const columnsStatesToUpdate = computeLineWrapStateUpdate(column, value);
        if (typeof oldValue === "boolean") {
          const columnIds = table.getAllLeafColumnsWithoutBuiltin().map((col) => col.id);
          const currentColumnsState = Object.fromEntries(columnIds.map((id) => [id, oldValue]));
          return {
            ...currentColumnsState,
            ...columnsStatesToUpdate
          };
        }
        return {
          ...oldValue,
          ...columnsStatesToUpdate
        };
      });
    };
    column.getIsLineWrapped = () => {
      const lineWrapState = table.getState().columnLineWrap;
      if (typeof lineWrapState === "boolean") {
        return lineWrapState;
      }
      const childColumns = column.columns;
      if (childColumns.length > 0) {
        return childColumns.every((child) => child.getIsLineWrapped());
      }
      return lineWrapState?.[column.id] ?? false;
    };
    column.getCanLineWrap = () => {
      return table.options.enableLineWrap ?? true;
    };
  },
  createTable(table) {
    table.setColumnLineWrap = (updater) => {
      const newValue = (0, import_react_table.functionalUpdate)(
        updater,
        table.getState().columnLineWrap ?? false
      );
      table.options.onColumnLineWrapChange?.(newValue);
    };
  }
};
function computeLineWrapStateUpdate(column, value) {
  const newValue = value ?? !column.getIsLineWrapped();
  const childColumns = column.columns;
  if (childColumns.length === 0) {
    return {
      [column.id]: newValue
    };
  }
  return Object.fromEntries(childColumns.map((child) => [child.id, newValue]));
}
function useColumnLineWrap(props, options) {
  const configLineWrap = (0, import_useDataTableConfig.useDataTableConfigElement)("lineWrap");
  const [columnLineWrap, onColumnLineWrapChange] = (0, import_useControllableState.useControllableState)({
    value: props.lineWrap ?? configLineWrap,
    defaultValue: props.defaultLineWrap,
    onChange: props.onLineWrapChange
  });
  options.state = { ...options.state, columnLineWrap };
  options.onColumnLineWrapChange = onColumnLineWrapChange;
}
function DataTableGlobalLineWrapToolbarAction() {
  const table = (0, import_DataTableContext.useDataTable)();
  const intl = (0, import_react_intl.useIntl)();
  const lineWrapState = table.getState().columnLineWrap;
  const Icon = lineWrapState === true ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.WrapTextOffIcon, {}) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.WrapTextIcon, {});
  const lineWrapLabel = lineWrapState === true ? intl.formatMessage(import_messages.dataTableMessages.disableLineWrapLabel) : intl.formatMessage(import_messages.dataTableMessages.enableLineWrapLabel);
  const lineWrapDtProperty = `state:${lineWrapState === true ? "unwrap" : "wrap"}`;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Tooltip.Tooltip, { text: lineWrapLabel, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Button.Button,
    {
      "aria-label": lineWrapLabel,
      "data-dt-name": "ToolbarLineWrap",
      "data-dt-properties": lineWrapDtProperty,
      onClick: () => table.setColumnLineWrap((oldLineWrap) => {
        if (oldLineWrap === true) {
          return {};
        }
        return true;
      }),
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: Icon })
    }
  ) });
}
