// @dynatrace/strato-components/charts/categorical-bar/components/toolbar/CategoricalBarChartToolbar.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CategoricalBarChartToolbar_exports = {};
__export(CategoricalBarChartToolbar_exports, {
  CategoricalBarChartToolbar: () => CategoricalBarChartToolbar
});
module.exports = __toCommonJS(CategoricalBarChartToolbar_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_DownloadDataButton = require("./DownloadDataButton.js");
var import_useBehavioralTrackingProps = require("../../../../core/hooks/useBehavioralTrackingProps.js");
var import_ChartToolbarRenderer = require("../../../core/components/toolbar/ChartToolbarRenderer.js");
var import_constants = require("../../constants.js");
var import_selectors = require("../../hooks/store/selectors.js");
var import_useCategoricalBarChartBaseConfig = require("../../hooks/useCategoricalBarChartBaseConfig.js");
var import_useCategoricalBarChartRawData = require("../../hooks/useCategoricalBarChartRawData.js");
var import_build_and_download_csv = require("../../utils/build-and-download-csv.js");
const CategoricalBarChartToolbar = ({
  onMouseEnter,
  onMouseLeave
}) => {
  const [isTooltipHidden, setIsTooltipHidden] = (0, import_react.useState)(false);
  const rawData = (0, import_useCategoricalBarChartRawData.useCategoricalBarChartRawData)();
  const { toolbar, intents } = (0, import_useCategoricalBarChartBaseConfig.useCategoricalBarChartBaseConfig)();
  const isInChartBounds = (0, import_selectors.useMouseInBoundsState)();
  const isDownloadEnabled = !!toolbar.downloadData?.enabled;
  const behaviorTrackingProps = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(toolbar)[0];
  if (toolbar.hidden || !(isDownloadEnabled || intents.length)) {
    return null;
  }
  const handleToggleMenu = (open) => {
    !open && onMouseLeave();
  };
  const handleDownloadDataToolbar = () => (0, import_build_and_download_csv.buildAndDownloadCsv)(rawData);
  const showDefaultToolbarDownloadButton = isDownloadEnabled && !intents.length;
  const showMenuToolbarButton = isDownloadEnabled && !!intents.length;
  const initialA11yState = showDefaultToolbarDownloadButton ? [{ download: 0 }] : [{ intent: 0 }];
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ChartToolbarRenderer.ChartToolbarRenderer,
    {
      visible: isInChartBounds,
      placement: toolbar.placement,
      collapsed: toolbar.collapsed,
      draggable: true,
      zIndex: import_constants.zIndicesCategoricalAreas.toolbar,
      initialA11yState,
      onToggleToolbar: setIsTooltipHidden,
      onToggleMenu: handleToggleMenu,
      onMouseEnter,
      onMouseLeave,
      menuDisplayMode: "never",
      intents,
      isDownloadEnabled: showMenuToolbarButton,
      handleDownloadDataToolbar,
      removeDragHandle: true,
      ...behaviorTrackingProps,
      children: showDefaultToolbarDownloadButton && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_DownloadDataButton.DownloadDataButton, { hideTooltip: isTooltipHidden })
    }
  );
};
