// @dynatrace/strato-components/esm/charts/meter-bar/multi-meter-bar/components/MultiMeterBarChartOverlay/MultiMeterBarChartOverlayWithTooltip.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import { useEffect, useState } from "react";
import { Flex } from "../../../../../layouts/flex/Flex.js";
import { interactionsOverlay } from "../../../../core/components/interactions-overlay/InteractionsOverlay.sty.js";
import { useTooltipData } from "../../../../core/components/overlay-tooltip/_to-be-removed/utils.js";
import { useOverlay } from "../../../../core/components/overlay-tooltip/contexts/OverlayContext.js";
import { OverlayTooltip } from "../../../../core/components/overlay-tooltip/OverlayTooltip.js";
import { isTooltipDataEmpty } from "../../../../core/components/overlay-tooltip/utils.js";
import { useSeriesActions } from "../../../../core/components/series-action/ChartSeriesActionContext.js";
import { MeterBarChartTooltip } from "../../../shared/components/meter-bar-tooltip/MeterBarChartTooltip.js";
import { useFormatter } from "../../../shared/context/FormatterContext.js";
import { useMeterBarLayoutSize } from "../../../shared/context/MeterBarLayoutSizeContext.js";
import { elementSize } from "../../../shared/styles/element-size.sty.js";
import { container } from "../../../shared/styles.sty.js";
import { getNumericValue } from "../../../shared/utils/get-numeric-value.js";
import { TOOLTIP_VARIANT } from "../../constants.js";
import { useMultiMeterBarChartTooltipData } from "../../context/MultiMeterBarChartTooltipContext.js";
import { useMultiMeterBarVisibleSegmentsState } from "../../hooks/store/selectors.js";
import { useMultiMeterBarChartTooltipHandlers } from "../../hooks/useMultiMeterBarChartTooltipHandlers.js";
const MultiMeterBarCustomTooltip = ({
  template,
  visibleSegments,
  hoveredIndex
}) => {
  const segments = visibleSegments.map((s) => ({
    name: s.name ?? "",
    value: getNumericValue(s.value, s.valueAccessor),
    color: s.color ?? ""
  }));
  const activeSegment = hoveredIndex !== void 0 && segments[hoveredIndex] ? { ...segments[hoveredIndex], index: hoveredIndex } : void 0;
  return /* @__PURE__ */ jsx(OverlayTooltip, { legacyTemplate: template({ segments, activeSegment }) });
};
const MultiMeterBarChartOverlayWithTooltip = (props) => {
  const [hoveredSegmentIndex, setHoveredSegmentIndex] = useState(void 0);
  const overlay = useOverlay();
  const { chartSeriesActions } = useSeriesActions();
  const { tooltipConfig } = useMultiMeterBarChartTooltipData();
  const { size } = useMeterBarLayoutSize();
  const { formatter, unit } = useFormatter();
  const visibleSegments = useMultiMeterBarVisibleSegmentsState();
  useEffect(() => {
    setHoveredSegmentIndex(void 0);
  }, [visibleSegments.length]);
  const isSharedTooltip = tooltipConfig.variant === TOOLTIP_VARIANT.Shared;
  const sharedData = {
    geometry: "multimeterbar",
    series: visibleSegments.map((item) => ({
      name: item.name ?? "",
      value: formatter(item.value, unit),
      color: item.color
    }))
  };
  const { onClick, onPointerMove, onPointerLeave } = useMultiMeterBarChartTooltipHandlers(sharedData, setHoveredSegmentIndex);
  const { data: tooltipData } = useTooltipData();
  const tooltipContent = tooltipConfig.template && !isTooltipDataEmpty(tooltipData) ? /* @__PURE__ */ jsx(
    MultiMeterBarCustomTooltip,
    {
      template: tooltipConfig.template,
      visibleSegments,
      hoveredIndex: hoveredSegmentIndex
    }
  ) : /* @__PURE__ */ jsx(
    MeterBarChartTooltip,
    {
      variant: isSharedTooltip ? TOOLTIP_VARIANT.Shared : TOOLTIP_VARIANT.Single,
      activeItem: hoveredSegmentIndex ?? -1,
      actionsForItem: chartSeriesActions ? (item) => () => chartSeriesActions(item) : void 0
    }
  );
  return /* @__PURE__ */ jsxs(
    Flex,
    {
      ref: overlay.overlayRef,
      className: clsx(container, elementSize({ size }), interactionsOverlay),
      "data-testid": "overlay-wrapper",
      onPointerMove,
      onPointerLeave,
      onClick,
      children: [
        props.children,
        tooltipContent
      ]
    }
  );
};
export {
  MultiMeterBarChartOverlayWithTooltip
};
