// @dynatrace/strato-components/esm/filters/timeframe-selector/components/presets/PresetItem.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import {
  useCallback,
  useEffect,
  useRef,
  useState
} from "react";
import { useIntl } from "react-intl";
import { HistoryIcon } from "@dynatrace/strato-icons";
import * as styles from "./Presets.sty.js";
import { useId } from "../../../../core/hooks/useId.js";
import { useFocusRing } from "../../../../core/styles/useFocusRing.js";
import { useTimeframeSelectorStore } from "../../../../core/time/timeframeStore/useTimeframeStore.js";
import { mergeProps } from "../../../../core/utils/merge-props.js";
import {
  parseValue,
  removePrecisionFromTimeValue
} from "../../../../forms/forms-core/time/timeParser.js";
import { Flex } from "../../../../layouts/flex/Flex.js";
import { Grid } from "../../../../layouts/grid/Grid.js";
import { Tooltip } from "../../../../overlays/tooltip/Tooltip.js";
import { Text } from "../../../../typography/text/Text.js";
import { TIMEFRAME_SELECTOR_MESSAGES } from "../../constants/timeframeSelectorMessages.js";
import { usePresetItemContext } from "../../hooks/usePresetItemContext.js";
import { useTimeframeSelectorContext } from "../../hooks/useTimeframeSelectorContext.js";
import { useTooltipMessages } from "../../hooks/useTooltipMessages.js";
import {
  isPresetSelected,
  saveRecentTimeframe
} from "../../utils/presets-helpers.js";
import { isOutsideRange } from "../../utils/timeframe-selection.js";
const PresetItem = (props) => {
  const {
    children,
    value,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    ...remainingProps
  } = props;
  const intl = useIntl();
  const [openTooltip, setOpenTooltip] = useState(false);
  const { details } = useTimeframeSelectorContext();
  const setValue = useTimeframeSelectorStore(
    (state2) => state2.setTimeframeSelectorTimeframe
  );
  const min = useTimeframeSelectorStore((state2) => state2.min?.value);
  const max = useTimeframeSelectorStore((state2) => state2.max?.value);
  const precision = useTimeframeSelectorStore((state2) => state2.precision);
  const { isRecent } = usePresetItemContext();
  const isSelected = isPresetSelected(value, details, isRecent, precision);
  const isDisabled = isOutsideRange(value, min, max);
  const tooltipId = useId();
  const tooltipFocusProps = {
    onFocus: useCallback(
      () => !isDisabled && setOpenTooltip(true),
      [isDisabled]
    ),
    onBlur: useCallback(() => setOpenTooltip(false), [])
  };
  const { focusProps, focusClassName } = useFocusRing({
    color: isSelected && !isDisabled ? "primary" : "neutral",
    disabled: isDisabled
  });
  let state = isSelected ? "selected" : "rest";
  if (isDisabled) {
    state = "disabled";
  }
  const {
    timeframeDisplayValue,
    formattedPresetItemValue,
    tooltipFormattedTimeValueStart,
    tooltipFormattedTimeValueEnd,
    isAbsoluteTimeframe,
    isInvalid
  } = useTooltipMessages(value, children);
  const handleItemClick = useCallback(() => {
    if (isDisabled) {
      return;
    }
    const { from, to } = parseValue(value);
    if (from && to) {
      const timeframe = {
        from: removePrecisionFromTimeValue(from, precision) ?? from,
        to: removePrecisionFromTimeValue(to, precision) ?? to
      };
      setValue(timeframe);
      saveRecentTimeframe(timeframe);
    }
  }, [isDisabled, value, precision, setValue]);
  const handleItemKeyDown = useCallback(
    (event) => {
      if (isDisabled) {
        return;
      }
      if (event.code === "Enter" || event.code === "NumpadEnter" || event.code === "Space") {
        handleItemClick();
      }
    },
    [handleItemClick, isDisabled]
  );
  const liRef = useRef(null);
  useEffect(() => {
    if (isSelected) {
      liRef.current?.scrollIntoView({ block: "nearest" });
    }
  }, []);
  if (isInvalid) {
    return null;
  }
  const tooltipContent = isDisabled ? /* @__PURE__ */ jsx(Text, { textStyle: "small", children: intl.formatMessage(TIMEFRAME_SELECTOR_MESSAGES.disabledTooltip) }) : /* @__PURE__ */ jsxs(Flex, { flexDirection: "column", children: [
    /* @__PURE__ */ jsxs(Flex, { gap: 0, flexDirection: "column", children: [
      /* @__PURE__ */ jsx(Text, { textStyle: "base-emphasized", children: timeframeDisplayValue }),
      /* @__PURE__ */ jsx(Text, { textStyle: "small", children: formattedPresetItemValue })
    ] }),
    !isAbsoluteTimeframe && /* @__PURE__ */ jsxs(Flex, { gap: 0, flexDirection: "column", children: [
      /* @__PURE__ */ jsx(Text, { textStyle: "small", children: tooltipFormattedTimeValueStart }),
      /* @__PURE__ */ jsx(Text, { textStyle: "small", children: tooltipFormattedTimeValueEnd })
    ] })
  ] });
  return /* @__PURE__ */ jsx(
    "li",
    {
      ref: liRef,
      role: "option",
      "aria-selected": isSelected,
      "aria-disabled": isDisabled,
      className: clsx([
        focusClassName,
        styles.item({
          state
        }),
        consumerClassName
      ]),
      onClick: handleItemClick,
      onKeyDown: handleItemKeyDown,
      "data-testid": dataTestId,
      style: consumerStyle,
      ...mergeProps(remainingProps, focusProps, tooltipFocusProps),
      children: /* @__PURE__ */ jsx(
        Tooltip,
        {
          id: tooltipId,
          text: tooltipContent,
          placement: "right",
          fallbackPlacements: [
            "right-end",
            "bottom-end",
            "bottom",
            "bottom-start",
            "left-end",
            "left",
            "left-start"
          ],
          open: openTooltip,
          onOpenChange: setOpenTooltip,
          children: /* @__PURE__ */ jsx(
            "div",
            {
              className: styles.wrapper({ isSelected, isDisabled }),
              "aria-describedby": tooltipId,
              children: /* @__PURE__ */ jsxs(Grid, { className: styles.itemContainer, children: [
                /* @__PURE__ */ jsxs(Flex, { alignItems: "center", className: styles.displayValueWrapper, children: [
                  isRecent && /* @__PURE__ */ jsx(HistoryIcon, { className: styles.recentIcon }),
                  /* @__PURE__ */ jsx(
                    Text,
                    {
                      textStyle: "base-emphasized",
                      as: "div",
                      className: styles.timeframeDisplayValue,
                      children: timeframeDisplayValue
                    }
                  )
                ] }),
                (!isRecent || !isAbsoluteTimeframe) && /* @__PURE__ */ jsx(
                  Text,
                  {
                    as: Flex,
                    textStyle: "small",
                    className: clsx(
                      styles.mediaQuery,
                      styles.formattedPresetValue({
                        state
                      })
                    ),
                    children: formattedPresetItemValue
                  }
                )
              ] })
            }
          )
        }
      )
    }
  );
};
PresetItem.displayName = "TimeframeSelector.PresetItem";
export {
  PresetItem
};
