// @dynatrace/strato-components/esm/filters/timeframe-selector/components/presets/Presets.sty.js
import "./Presets.css";
import { createRuntimeFn as _7a468 } from "@vanilla-extract/recipes/createRuntimeFn";
var displayValueWrapper = "_14rlfn36-3-6-0";
var formattedPresetValue = _7a468({ defaultClassName: "_14rlfn3e-3-6-0", variantClassNames: { state: { rest: "_14rlfn3f-3-6-0", selected: "_14rlfn3g-3-6-0", disabled: "_14rlfn3h-3-6-0" } }, defaultVariants: { state: "rest" }, compoundVariants: [] });
var indicatorIcon = _7a468({ defaultClassName: "_14rlfn3i-3-6-0", variantClassNames: { isDisabled: { true: "_14rlfn3j-3-6-0", false: "_14rlfn3k-3-6-0" } }, defaultVariants: {}, compoundVariants: [] });
var item = _7a468({ defaultClassName: "_14rlfn31-3-6-0", variantClassNames: { state: { rest: "_14rlfn32-3-6-0", selected: "_14rlfn33-3-6-0", disabled: "_14rlfn34-3-6-0" }, defaultVariants: { state: "rest" } }, defaultVariants: {}, compoundVariants: [] });
var itemContainer = "_14rlfn30-3-6-0";
var listContainer = "_14rlfn3l-3-6-0";
var mediaQuery = "_14rlfn3d-3-6-0";
var recentHeading = "_14rlfn3m-3-6-0";
var recentIcon = "_14rlfn3n-3-6-0";
var timeframeDisplayValue = "_14rlfn37-3-6-0";
var wrapper = _7a468({ defaultClassName: "_14rlfn38-3-6-0", variantClassNames: { isSelected: { true: "_14rlfn39-3-6-0", false: "_14rlfn3a-3-6-0" }, isDisabled: { true: "_14rlfn3b-3-6-0", false: "_14rlfn3c-3-6-0" } }, defaultVariants: {}, compoundVariants: [] });
export {
  displayValueWrapper,
  formattedPresetValue,
  indicatorIcon,
  item,
  itemContainer,
  listContainer,
  mediaQuery,
  recentHeading,
  recentIcon,
  timeframeDisplayValue,
  wrapper
};
