// @dynatrace/strato-components/esm/filters/timeframe-selector/components/presets/Presets.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import { Children } from "react";
import { useIntl } from "react-intl";
import { PresetItem } from "./PresetItem.js";
import * as styles from "./Presets.sty.js";
import { isSlot } from "../../../../core/utils/is-slot.js";
import { devLog } from "../../../../core/utils/logging.js";
import { TIMEFRAME_SELECTOR_PRESETS } from "../../constants/timeframe-presets.js";
import { TIMEFRAME_SELECTOR_MESSAGES } from "../../constants/timeframeSelectorMessages.js";
function getItemSlots(children) {
  if (!children) {
    return null;
  }
  const validChildren = [];
  for (const child of Children.toArray(children)) {
    if (isSlot(child, PresetItem)) {
      validChildren.push(child);
    } else {
      devLog(
        "The used child is not supported by the TimeframeSelector.Presets."
      );
    }
  }
  return validChildren.length > 0 ? validChildren : null;
}
const Presets = (props) => {
  const {
    children,
    "aria-label": ariaLabel,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    ...remainingProps
  } = props;
  const intl = useIntl();
  const presetsListLabel = ariaLabel ?? intl.formatMessage(TIMEFRAME_SELECTOR_MESSAGES.timeframePresets);
  const validChildren = getItemSlots(children);
  return /* @__PURE__ */ jsx(
    "ul",
    {
      role: "listbox",
      "aria-label": presetsListLabel,
      className: clsx([styles.listContainer, consumerClassName]),
      style: consumerStyle,
      "data-testid": dataTestId,
      ...remainingProps,
      children: validChildren ?? TIMEFRAME_SELECTOR_PRESETS.map((item) => /* @__PURE__ */ jsx(
        PresetItem,
        {
          value: item,
          "data-dt-name": "Preset",
          "data-dt-properties": `valueFrom:${item.from},valueTo:${item.to}`
        },
        `${item.from} \u2192 ${item.to}`
      ))
    }
  );
};
Presets["displayName"] = "TimeframeSelector.Presets";
export {
  Presets
};
