// @dynatrace/strato-components/esm/filters/timeframe-selector/components/presets/RecentTimeframes.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { useQuery } from "@tanstack/react-query";
import { useMemo } from "react";
import { useIntl } from "react-intl";
import { PresetItem } from "./PresetItem.js";
import * as styles from "./Presets.sty.js";
import { Divider } from "../../../../layouts/divider/Divider.js";
import { Text } from "../../../../typography/text/Text.js";
import { TIMEFRAME_SELECTOR_MESSAGES } from "../../constants/timeframeSelectorMessages.js";
import { PresetItemContext } from "../../contexts/PresetItemContext.js";
import { getTimeframePresets } from "../../utils/presets-helpers.js";
import {
  timeframeQueryClient,
  TIMEFRAME_PRESETS_QUERY_KEY
} from "../../utils/TimeframePresetsClient.js";
const RecentTimeframes = () => {
  const intl = useIntl();
  const { data, isLoading } = useQuery(
    {
      queryKey: TIMEFRAME_PRESETS_QUERY_KEY,
      queryFn: getTimeframePresets,
      notifyOnChangeProps: ["dataUpdatedAt"]
    },
    timeframeQueryClient
  );
  const presetItemContextValue = useMemo(
    () => ({ isRecent: true }),
    []
  );
  if (!data?.recentPresets?.length || isLoading) {
    return null;
  }
  const ariaLabel = intl.formatMessage(
    TIMEFRAME_SELECTOR_MESSAGES.recentTimeframesList
  );
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(Divider, { flexItem: false }),
    /* @__PURE__ */ jsxs("div", { children: [
      /* @__PURE__ */ jsx(Text, { textStyle: "small", className: styles.recentHeading, as: "div", children: intl.formatMessage(
        TIMEFRAME_SELECTOR_MESSAGES.recentTimeframesHeading
      ) }),
      /* @__PURE__ */ jsx(
        "ul",
        {
          role: "listbox",
          className: styles.listContainer,
          "aria-label": ariaLabel,
          children: /* @__PURE__ */ jsx(PresetItemContext.Provider, { value: presetItemContextValue, children: data.recentPresets.map(({ timeframe }, index) => /* @__PURE__ */ jsx(
            PresetItem,
            {
              value: timeframe,
              "data-dt-name": "Recent",
              "data-dt-properties": `index:${index},valueFrom:${timeframe.from},valueTo:${timeframe.to}`
            },
            `${timeframe.from} \u2192 ${timeframe.to}`
          )) })
        }
      )
    ] })
  ] });
};
RecentTimeframes.displayName = "RecentTimeframes";
export {
  RecentTimeframes
};
