// @dynatrace-sdk/error-handlers/cjs/index.js
/**
 * @license
 * Copyright 2023 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/platform/error-handlers/src/index.ts
var src_exports = {};
__export(src_exports, {
  ErrorType: () => ErrorType,
  addGlobalErrorSerializer: () => addGlobalErrorSerializer,
  isGlobalErrorSerializerSupported: () => isGlobalErrorSerializerSupported
});
module.exports = __toCommonJS(src_exports);

// packages/platform/shared/utils/get-global-with-dt-runtime.ts
var getGlobalWithDtRuntime = () => typeof globalThis !== "undefined" ? globalThis : window;

// packages/platform/error-handlers/src/lib/add-global-error-serializer.ts
function addGlobalErrorSerializer(serializer) {
  const addGlobalErrorSerializerFunc = getAddGlobalErrorSerializer();
  if (typeof addGlobalErrorSerializerFunc === "function") {
    addGlobalErrorSerializerFunc(serializer);
  } else {
    console.warn("Missing addGlobalErrorSerializer function from web runtime.");
  }
}
function isGlobalErrorSerializerSupported() {
  return typeof getAddGlobalErrorSerializer() === "function";
}
function getAddGlobalErrorSerializer() {
  return getGlobalWithDtRuntime().dtRuntime?.errorHandlers?.addGlobalErrorSerializer;
}

// packages/platform/error-handlers/src/lib/types/error-type.ts
var ErrorType = /* @__PURE__ */ ((ErrorType2) => {
  ErrorType2["COMMON"] = "JS Error";
  ErrorType2["HTTP"] = "Http Error";
  return ErrorType2;
})(ErrorType || {});
