// @dynatrace/strato-components/charts/meter-bar/shared/components/meter-bar-toolbar/MeterBarToolbar.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MeterBarToolbar_exports = {};
__export(MeterBarToolbar_exports, {
  MeterBarChartToolbar: () => MeterBarChartToolbar
});
module.exports = __toCommonJS(MeterBarToolbar_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_useBehavioralTrackingProps = require("../../../../../core/hooks/useBehavioralTrackingProps.js");
var import_useTooltipActions = require("../../../../core/components/overlay-tooltip/useTooltipActions.js");
var import_ChartToolbarRenderer = require("../../../../core/components/toolbar/ChartToolbarRenderer.js");
var import_zoom = require("../../../../core/utils/zoom.js");
var import_MeterBarToolbarConfigContext = require("../../context/MeterBarToolbarConfigContext.js");
const MeterBarChartToolbar = () => {
  const toolbarConfig = (0, import_MeterBarToolbarConfigContext.useMeterBarToolbarConfig)();
  const isToolbarEnabled = !toolbarConfig?.hidden && !!toolbarConfig?.intents?.length;
  const behaviorTrackingProps = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(
    toolbarConfig ?? {}
  )[0];
  const { reset } = (0, import_useTooltipActions.useTooltipActions)();
  if (!isToolbarEnabled) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ChartToolbarRenderer.ChartToolbarRenderer,
    {
      visible: isToolbarEnabled,
      placement: toolbarConfig.placement,
      collapsed: toolbarConfig.collapsed,
      draggable: false,
      initialA11yState: (0, import_zoom.getA11yToolbarConfig)({
        intent: !!toolbarConfig.intents
      }),
      menuDisplayMode: "never",
      intents: toolbarConfig.intents,
      isDownloadEnabled: false,
      toolbarDisplayMode: "inline",
      overlayClassName: "",
      onToggleMenu: (isOpen) => {
        if (isOpen) {
          reset();
        }
      },
      ...behaviorTrackingProps
    }
  );
};
MeterBarChartToolbar["displayName"] = "MeterBarChartToolbar";
