// @dynatrace/strato-components/esm/charts/core/components/annotations/components/displays/AnnotationsNumericalDisplay.js
import { jsx, jsxs } from "react/jsx-runtime";
import { AnnotationsContainer } from "./AnnotationsContainer.js";
import { useBehavioralTrackingProps } from "../../../../../../core/hooks/useBehavioralTrackingProps.js";
import { OverlayProvider } from "../../../overlay-tooltip/contexts/OverlayContext.js";
import { TooltipProvider } from "../../../overlay-tooltip/contexts/TooltipContext.js";
import { AnnotationsProviders } from "../../providers/AnnotationsProviders.js";
import { AnnotationsRenderer } from "../AnnotationsRenderer.js";
import { AnnotationsOverlayTooltip } from "../tooltip/AnnotationsOverlayTooltip.js";
import { AnnotationsTooltipFormatterContext } from "../tooltip/contexts/AnnotationsTooltipFormatter.context.js";
const AnnotationsNumericalDisplay = (props) => {
  const {
    annotationsActions,
    markerGroupHover,
    formatter,
    tooltipConfig,
    "data-dt-component": dataDtComponent,
    ...rendererProps
  } = props;
  const [behaviorTrackingProps] = useBehavioralTrackingProps(props);
  return /* @__PURE__ */ jsx(OverlayProvider, { children: /* @__PURE__ */ jsx(
    AnnotationsContainer,
    {
      ...behaviorTrackingProps,
      "data-dt-component": dataDtComponent,
      children: /* @__PURE__ */ jsx(TooltipProvider, { persistedStore: true, children: /* @__PURE__ */ jsx(
        AnnotationsProviders,
        {
          onAnnotationsMarkersInteraction: markerGroupHover,
          annotationsActions,
          tooltipConfig,
          children: /* @__PURE__ */ jsxs(AnnotationsTooltipFormatterContext.Provider, { value: formatter, children: [
            /* @__PURE__ */ jsx(AnnotationsRenderer, { ...rendererProps }),
            /* @__PURE__ */ jsx(
              AnnotationsOverlayTooltip,
              {
                tooltipType: "numerical",
                truncationMode: rendererProps.truncateMode,
                maxLines: rendererProps.maxLines
              }
            )
          ] })
        }
      ) })
    }
  ) });
};
AnnotationsNumericalDisplay["displayName"] = "AnnotationsNumericalDisplay";
export {
  AnnotationsNumericalDisplay
};
