// @dynatrace/strato-components/charts/core/components/single-value-renderer/components/trend/Trend.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Trend_exports = {};
__export(Trend_exports, {
  Trend: () => Trend
});
module.exports = __toCommonJS(Trend_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_Icon = require("./components/Icon.js");
var import_Trend_css = require("./Trend.sty.js");
var import_define_trend_colors = require("./utils/define-trend-colors.js");
var import_define_trend_direction = require("./utils/define-trend-direction.js");
var import_useId = require("../../../../../../core/hooks/useId.js");
var import_useMergeRefs = require("../../../../../../core/hooks/useMergeRefs.js");
var import_Flex = require("../../../../../../layouts/flex/Flex.js");
var import_useFontSizeForContainer = require("../../../../hooks/useFontSizeForContainer.js");
var import_useThrottledResizeObserver = require("../../../../hooks/useThrottledResizeObserver.js");
var import_typographyStyles_css = require("../../../../typographyStyles.sty.js");
var import_get_formatter = require("../../../../utils/get-formatter.js");
var import_constants = require("../../constants.js");
var import_single_value_context = require("../../contexts/single-value.context.js");
var import_use_global_font_size = require("../../hooks/use-global-font-size.js");
const Trend = (props) => {
  const {
    direction: directionProp,
    showIcon = true,
    value: valueProp,
    label,
    inverseTrend = "none",
    colorsOverride,
    formatter: formatterProp,
    contrastColor
  } = props;
  const { alignment } = (0, import_react.useContext)(import_single_value_context.SingleValueContext);
  const direction = (0, import_define_trend_direction.defineTrendDirection)(directionProp, valueProp);
  const [iconColor, valueColor] = (0, import_define_trend_colors.defineTrendColors)(
    direction,
    inverseTrend,
    colorsOverride
  );
  const formatter = (0, import_get_formatter.getFormatter)(formatterProp);
  const formattedValue = !(0, import_lodash_es.isNil)(valueProp) ? formatter(valueProp) : "";
  const value = formattedValue.replace(" ", "");
  const ref = (0, import_react.useRef)(null);
  const { refCallback, width = 0, height = 0 } = (0, import_useThrottledResizeObserver.useThrottledResizeObserver)();
  const mergedRef = (0, import_useMergeRefs.useMergeRefs)([refCallback, ref]);
  const showTrendArrow = showIcon && (!(0, import_lodash_es.isEmpty)(value) || !(0, import_lodash_es.isEmpty)(label));
  const preferredFontSize = (0, import_useFontSizeForContainer.useFontSizeForContainer)({
    container: { width, height: height * import_constants.FONT_HEIGHT_RATIO },
    text: showIcon ? `00${value}0${label}` : `${value}0${label}`,
    element: ref.current
  });
  const id = (0, import_useId.useId)();
  const { updateTrendFontSize } = (0, import_use_global_font_size.useGlobalFontSize)();
  (0, import_react.useLayoutEffect)(() => {
    if (height && width) {
      updateTrendFontSize?.(id, preferredFontSize);
    }
  }, [height, width, preferredFontSize]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_Flex.Flex,
    {
      className: import_Trend_css.trendContainer,
      style: {
        [import_constants.TREND_FONT_CSS_TOKEN]: preferredFontSize
      },
      ref: mergedRef,
      "data-testid": "single-value-trend",
      alignItems: "center",
      justifyContent: alignment,
      gap: 0,
      children: [
        showTrendArrow && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Icon.Icon, { direction, color: contrastColor ?? iconColor }),
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { className: import_Trend_css.valueLabelContainer, children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "span",
            {
              className: import_typographyStyles_css.replaceSlashedZero,
              style: {
                color: contrastColor ?? valueColor,
                paddingLeft: showTrendArrow ? "2ch" : "0"
              },
              children: value
            }
          ),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "span",
            {
              className: import_Trend_css.trendLabel,
              style: {
                color: contrastColor ?? import_constants.TREND_LABEL_DEFAULT_COLOR
              },
              children: label
            }
          )
        ] })
      ]
    }
  );
};
Trend["displayName"] = "SingleValue.Trend";
