// @dynatrace/strato-components/tables/DataTable/components/Toolbar/DataTableToolbarMenu.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DataTableToolbarMenu_exports = {};
__export(DataTableToolbarMenu_exports, {
  DataTableToolbarMenu: () => DataTableToolbarMenu
});
module.exports = __toCommonJS(DataTableToolbarMenu_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_DataTableDownloadData = require("./Download/DataTableDownloadData.js");
var import_Button = require("../../../../buttons/button/Button.js");
var import_Menu = require("../../../../navigation/menu/Menu.js");
var import_Tooltip = require("../../../../overlays/tooltip/Tooltip.js");
var import_messages = require("../../messages.js");
const DataTableToolbarMenu = (props) => {
  const { hasDownloadData, downloadExcludedColumns, intents } = props;
  const [menuOpen, setMenuOpen] = (0, import_react.useState)(false);
  const [isDownloading, setIsDownloading] = (0, import_react.useState)(false);
  const hasIntents = intents.length > 0;
  const triggerIcon = hasIntents ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.DotMenuIcon, {}) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.DownloadIcon, {});
  const Group = hasIntents ? import_Menu.Menu.Group : import_react.Fragment;
  const intl = (0, import_react_intl.useIntl)();
  let toolbarMenuMessage = import_messages.dataTableMessages.downloadButtonLabel;
  if (isDownloading) {
    toolbarMenuMessage = import_messages.dataTableMessages.downloadingButtonLabel;
  } else if (hasIntents) {
    toolbarMenuMessage = import_messages.dataTableMessages.toolbarButtonLabel;
  }
  const toolbarMenuLabel = intl.formatMessage(toolbarMenuMessage);
  const simpleButtonRef = (0, import_react.useRef)(null);
  if (!hasDownloadData && !hasIntents) {
    return;
  }
  if (menuOpen === false) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Tooltip.Tooltip, { text: toolbarMenuLabel, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Button.Button,
      {
        ref: simpleButtonRef,
        variant: "default",
        "aria-label": toolbarMenuLabel,
        onClick: () => setMenuOpen(true),
        loading: isDownloading,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: triggerIcon })
      }
    ) });
  }
  const {
    intentsWithAppOrIntentId: intentsWithAppId,
    intentsWithoutAppOrIntentId: intentsWithoutAppId
  } = intents.reduce(
    (acc, intentProps, idx) => {
      const hasAppOrIntentId = intentProps.options.recommendedAppId !== void 0 || intentProps.options.recommendedIntentId !== void 0;
      const intent = /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Menu.Menu.Intent, { ...intentProps, children: [
        intentProps.icon && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Prefix, { children: intentProps.icon }),
        intentProps.children
      ] }, `intent-${idx}`);
      hasAppOrIntentId ? acc.intentsWithAppOrIntentId.push(intent) : acc.intentsWithoutAppOrIntentId.push(intent);
      return acc;
    },
    { intentsWithAppOrIntentId: [], intentsWithoutAppOrIntentId: [] }
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_Menu.Menu,
    {
      open: menuOpen,
      onOpenChange: (open) => {
        if (open === false) {
          requestAnimationFrame(() => {
            simpleButtonRef.current?.focus();
          });
        }
        setMenuOpen(open);
      },
      modal: false,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Tooltip.Tooltip, { text: toolbarMenuLabel, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Trigger, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button, { variant: "default", "aria-label": toolbarMenuLabel, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: triggerIcon }) }) }) }),
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Menu.Menu.Content, { alignment: "end", children: [
          hasDownloadData && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(Group, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_DataTableDownloadData.DataTableDownloadData,
            {
              excludeColumns: downloadExcludedColumns,
              hasIntents,
              isDownloading,
              onDownloadStart: () => setIsDownloading(true),
              onDownloadEnd: () => setIsDownloading(false)
            }
          ) }),
          intentsWithAppId.length > 0 && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(Group, { children: intentsWithAppId }),
          intentsWithoutAppId.length > 0 && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(Group, { children: intentsWithoutAppId })
        ] })
      ]
    }
  );
};
