// @dynatrace/strato-components/tables/DataTable/components/Toolbar/DataTableToolbar.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DataTableToolbar_exports = {};
__export(DataTableToolbar_exports, {
  DataTableToolbar: () => DataTableToolbar,
  _isDataTableToolbar: () => _isDataTableToolbar,
  getToolbarConfig: () => getToolbarConfig
});
module.exports = __toCommonJS(DataTableToolbar_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_DataTableColumnOrderSettings = require("./ColumnSettings/DataTableColumnOrderSettings.js");
var import_DataTableColumnPinningSettings = require("./ColumnSettings/DataTableColumnPinningSettings.js");
var import_DataTableColumnSettings = require("./ColumnSettings/DataTableColumnSettings.js");
var import_DataTableColumnSettingsTrigger = require("./ColumnSettings/DataTableColumnSettingsTrigger.js");
var import_DataTableColumnVisibilitySettings = require("./ColumnSettings/DataTableColumnVisibilitySettings.js");
var import_DataTableToolbarMenu = require("./DataTableToolbarMenu.js");
var import_useBehavioralTrackingProps = require("../../../../core/hooks/useBehavioralTrackingProps.js");
var import_Intent = require("../../../../core/slots/Intent/Intent.js");
var import_is_slot = require("../../../../core/utils/is-slot.js");
var import_logging = require("../../../../core/utils/logging.js");
var import_Flex = require("../../../../layouts/flex/Flex.js");
var import_Download = require("../../features/Download/Download.js");
var import_LineWrap = require("../../features/LineWrap/LineWrap.js");
function _isDataTableToolbar(child) {
  return child?.type === DataTableToolbar;
}
function getToolbarConfig(children) {
  const slots = {
    visibilitySettings: false,
    columnOrderSettings: false,
    columnPinningSettings: false,
    downloadData: false,
    globalLineWrap: false
  };
  const toolbarItemProps = { intents: [] };
  import_react.Children.forEach(children, (child) => {
    if ((0, import_is_slot.isSlot)(child, import_DataTableColumnVisibilitySettings.DataTableColumnVisibilitySettings)) {
      slots.visibilitySettings = true;
      toolbarItemProps.resetColumnVisibility = child.props?.resetColumnVisibility;
    } else if ((0, import_is_slot.isSlot)(child, import_DataTableColumnOrderSettings.DataTableColumnOrderSettings)) {
      slots.columnOrderSettings = true;
      toolbarItemProps.resetColumnOrder = child.props?.resetColumnOrder;
    } else if ((0, import_is_slot.isSlot)(child, import_DataTableColumnPinningSettings.DataTableColumnPinningSettings)) {
      slots.columnPinningSettings = true;
      toolbarItemProps.resetColumnPinning = child.props?.resetColumnPinning;
    } else if ((0, import_is_slot.isSlot)(child, import_DataTableColumnSettingsTrigger.DataTableColumnSettingsTrigger)) {
      toolbarItemProps.customTriggerProps = child.props;
    } else if ((0, import_is_slot.isSlot)(child, import_Download.DataTableDownload)) {
      slots.downloadData = true;
      toolbarItemProps.downloadExcludedColumns = child.props?.excludeColumns;
      toolbarItemProps.onDownloadData = child.props?.onDownloadData;
    } else if ((0, import_is_slot.isSlot)(child, import_LineWrap.DataTableGlobalLineWrapAction)) {
      slots.globalLineWrap = true;
    } else if ((0, import_is_slot.isSlot)(child, import_Intent.Intent)) {
      toolbarItemProps.intents.push(child.props);
    } else {
      (0, import_logging.devLog)("Invalid child of DataTableToolbar");
    }
  });
  return { slots, toolbarItemProps };
}
const DataTableToolbar = (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const {
      children,
      "data-testid": dataTestId,
      className: consumerClassName,
      style: consumerStyle,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
    const { slots, toolbarItemProps } = getToolbarConfig(children);
    if (import_react.Children.count(children) === 0) {
      return null;
    }
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      import_Flex.Flex,
      {
        ref: forwardedRef,
        justifyContent: "flex-end",
        alignItems: "flex-end",
        gap: 4,
        "data-testid": dataTestId,
        className: consumerClassName,
        style: {
          gridArea: "1 / 2 / 2 / 2",
          ...consumerStyle
        },
        ...behavioralTrackingProps,
        ...restProps,
        children: [
          (slots.visibilitySettings || slots.columnOrderSettings || slots.columnPinningSettings) && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_DataTableColumnSettings.DataTableColumnSettings,
            {
              customTriggerProps: toolbarItemProps?.customTriggerProps
            }
          ),
          slots.globalLineWrap && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_LineWrap.DataTableGlobalLineWrapToolbarAction, {}),
          (slots.downloadData || toolbarItemProps.intents.length > 0) && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_DataTableToolbarMenu.DataTableToolbarMenu,
            {
              hasDownloadData: slots.downloadData,
              downloadExcludedColumns: toolbarItemProps?.downloadExcludedColumns,
              intents: toolbarItemProps.intents
            }
          )
        ]
      }
    );
  }
);
DataTableToolbar.displayName = "DataTable.Toolbar";
