// @dynatrace/strato-components/esm/core/time/helpers.js
import { TZDateMini } from "@date-fns/tz";
import { isSameDay, isValid } from "date-fns";
import { formatDate, parseTimeAsTimeValue } from "@dynatrace-sdk/units";
import { getTimezone } from "@dynatrace-sdk/user-preferences";
function isEqualTimeValue(tv1, tv2) {
  if (tv1 === tv2) {
    return true;
  }
  if (typeof tv1 === "string" || typeof tv2 === "string") {
    return false;
  }
  if (tv1 === null || tv2 === null) {
    return false;
  }
  if (tv1?.type !== tv2?.type) {
    return false;
  }
  return tv1?.value === tv2?.value;
}
function isEqualRangeSelection(tf1, tf2) {
  if (tf1 === tf2) {
    return true;
  }
  return isEqualTimeValue(tf1?.from ?? null, tf2?.from ?? null) && isEqualTimeValue(tf1?.to ?? null, tf2?.to ?? null);
}
function isSimpleTimeOffset(offset) {
  return typeof offset !== "object";
}
function isTimeframe(date, type) {
  return type === "range";
}
function isTimeValue(value) {
  return typeof value === "object" && value !== null && "absoluteDate" in value && "value" in value && "type" in value;
}
function isFullTimeframe(timeframe) {
  if (typeof timeframe !== "object" || timeframe === null) {
    return false;
  }
  const candidate = timeframe;
  return isTimeValue(candidate.from) && candidate.from.absoluteDate !== "" && isTimeValue(candidate.to) && candidate.to.absoluteDate !== "";
}
function isPartialTimeframe(timeframe) {
  if (typeof timeframe !== "object" || timeframe === null) {
    return false;
  }
  const candidate = timeframe;
  const hasFrom = "from" in candidate && candidate.from !== null;
  const hasTo = "to" in candidate && candidate.to !== null;
  return hasFrom !== hasTo;
}
function getFormattedTimeValue(timeValue, precision, showTimeOnly) {
  if (!timeValue) {
    return "";
  }
  const timezone = getTimezone();
  const formatOptions = {
    // Extract the user-settings timezone to use for the calculation
    timeZone: getTimezone(),
    day: showTimeOnly ? void 0 : "numeric",
    month: showTimeOnly ? void 0 : "short",
    year: showTimeOnly ? void 0 : "numeric",
    hour: "numeric",
    minute: "numeric",
    second: precision === "minutes" ? void 0 : "numeric",
    fractionalSecondDigits: precision === "milliseconds" ? 3 : void 0
  };
  const absoluteDate = timeValue.absoluteDate ? new TZDateMini(timeValue.absoluteDate, timezone) : null;
  return absoluteDate && isValid(absoluteDate) ? formatDate(absoluteDate, formatOptions) : "";
}
function getFormattedMinMax(min, max, precision) {
  const relativeNow = TZDateMini.tz(getTimezone()).getTime();
  const minDate = min ? parseTimeAsTimeValue(min, relativeNow) : null;
  const maxDate = max ? parseTimeAsTimeValue(max, relativeNow) : null;
  const newMinDate = new TZDateMini(minDate?.absoluteDate ?? "", getTimezone());
  if (!isNaN(newMinDate.getTime())) {
    if (precision === "seconds" && newMinDate.getMilliseconds() > 0) {
      newMinDate.setSeconds(newMinDate.getSeconds() + 1);
    } else if (precision === "minutes" && (newMinDate.getMilliseconds() > 0 || newMinDate.getSeconds() > 0)) {
      newMinDate.setMinutes(newMinDate.getMinutes() + 1);
    }
  }
  const minTimeValue = parseTimeAsTimeValue(
    !isNaN(newMinDate.getTime()) ? newMinDate.toISOString() : minDate?.absoluteDate,
    relativeNow
  );
  const formattedMin = getFormattedTimeValue(minTimeValue, precision);
  const formattedMax = getFormattedTimeValue(
    maxDate,
    precision,
    !!minTimeValue && !!maxDate && isSameDay(minTimeValue.absoluteDate, maxDate.absoluteDate)
  );
  return {
    min: formattedMin,
    max: formattedMax
  };
}
function mergeRangeValue(current, patch) {
  const base = current ?? { from: null, to: null };
  return {
    from: "from" in patch ? patch.from ?? null : base.from,
    to: "to" in patch ? patch.to ?? null : base.to
  };
}
export {
  getFormattedMinMax,
  getFormattedTimeValue,
  isEqualRangeSelection,
  isEqualTimeValue,
  isFullTimeframe,
  isPartialTimeframe,
  isSimpleTimeOffset,
  isTimeValue,
  isTimeframe,
  mergeRangeValue
};
