// @dynatrace/strato-components/esm/core/time/useDateTimeMessages.js
import { TZDateMini } from "@date-fns/tz";
import { useMemo } from "react";
import {
  createIntl,
  defineMessages,
  useIntl
} from "react-intl";
import { getTimezone } from "@dynatrace-sdk/user-preferences";
import { intlCache } from "../../forms/date-time-picker/utils/intl-cache.js";
import {
  getDayPeriodValues,
  getSafeRegionalFormat
} from "../../forms/forms-core/time/helpers.js";
const placeholderMessages = defineMessages({
  year: {
    defaultMessage: "YYYY",
    id: "KgAwef2nXv2+/xgE",
    description: "The abbreviation of the year part of a date value"
  },
  month: {
    defaultMessage: "MM",
    id: "Bo8Z1UaJVL3ONFo8",
    description: "The abbreviation of the month part of a date value"
  },
  day: {
    defaultMessage: "DD",
    id: "42g95tLLPXcqbFLl",
    description: "The abbreviation of the day part of a date value"
  },
  hour: {
    defaultMessage: "hh",
    id: "5+yhrNhI2AlQmDAo",
    description: "The abbreviation of the hour part of a time value"
  },
  minute: {
    defaultMessage: "mm",
    id: "Xup0Ti6gaxC1KDSY",
    description: "The abbreviation of the minute part of a time value"
  },
  second: {
    defaultMessage: "ss",
    id: "ktoTdX660gW5R80Z",
    description: "The abbreviation of the second part of a time value"
  },
  fractionalSecond: {
    defaultMessage: "sss",
    id: "ERAhDK0Wy/n7Sd2e",
    description: "The abbreviation of the fractional of the second part of a time value"
  }
});
function getPlaceholder(type, formatMessage) {
  const messageDescriptor = placeholderMessages[type];
  if (messageDescriptor) {
    return formatMessage(messageDescriptor);
  }
  if (type === "dayPeriod") {
    return getDayPeriodValues().join("/");
  }
  return type;
}
const useDateTimeMessages = (options) => {
  const {
    locale = getSafeRegionalFormat(),
    timeZone: timeZoneProp = getTimezone(),
    type = "datetime",
    precision = "minutes"
  } = options ?? {};
  const { formatMessage } = useIntl();
  const { formatDateToParts, timeZone } = useMemo(
    () => createIntl({ locale, timeZone: timeZoneProp }, intlCache),
    [locale, timeZoneProp]
  );
  const formatOptions = useMemo(() => {
    const options2 = {};
    switch (type) {
      case "date":
        options2["year"] = "numeric";
        options2["month"] = "2-digit";
        options2["day"] = "2-digit";
        return options2;
      // Return config and ignore precision prop if only the date is shown.
      case "datetime":
        options2["year"] = "numeric";
        options2["month"] = "2-digit";
        options2["day"] = "2-digit";
        options2["hour"] = "2-digit";
        options2["minute"] = "2-digit";
        break;
      case "time":
        options2["hour"] = "numeric";
        options2["minute"] = "2-digit";
        break;
    }
    switch (precision) {
      case "seconds":
        options2["second"] = "2-digit";
        break;
      case "milliseconds":
        options2["second"] = "2-digit";
        options2["fractionalSecondDigits"] = 3;
        break;
    }
    return options2;
  }, [type, precision]);
  const parts = useMemo(
    () => formatDateToParts(TZDateMini.tz(getTimezone()).getTime(), formatOptions),
    [formatOptions, formatDateToParts]
  );
  const formatString = useMemo(
    () => parts.map(
      ({ type: type2, value }) => type2 === "literal" ? value : getPlaceholder(type2, formatMessage)
    ).join(""),
    [formatMessage, parts]
  );
  return {
    locale,
    timeZone,
    formatDateToParts,
    parts,
    formatOptions,
    formatString,
    getPlaceholder
  };
};
export {
  getPlaceholder,
  useDateTimeMessages
};
