// @dynatrace-sdk/app-environment/esm/index.js
/**
 * @license
 * Copyright 2023-2025 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// packages/platform/shared/utils/get-global-with-dt-runtime.ts
var getGlobalWithDtRuntime = () => typeof globalThis !== "undefined" ? globalThis : window;

// packages/platform/app-environment/src/lib/get-app-id.ts
function getAppId() {
  const runtimeFunc = getGlobalWithDtRuntime().dtRuntime?.appEnvironment?.getAppId;
  if (typeof runtimeFunc === "function") {
    return runtimeFunc();
  } else {
    console.warn("Missing getAppId function from web runtime.");
    return "dt.missing.app.id";
  }
}

// packages/platform/app-environment/src/lib/get-app-name.ts
function getAppName() {
  const runtimeFunc = getGlobalWithDtRuntime().dtRuntime?.appEnvironment?.getAppName;
  if (typeof runtimeFunc === "function") {
    return runtimeFunc();
  } else {
    console.warn("Missing getAppName function from web runtime.");
    return "dt.missing.app.name";
  }
}

// packages/platform/app-environment/src/lib/get-app-version.ts
function getAppVersion() {
  const runtimeFunc = getGlobalWithDtRuntime().dtRuntime?.appEnvironment?.getAppVersion;
  if (typeof runtimeFunc === "function") {
    return runtimeFunc();
  } else {
    console.warn("Missing getAppVersion function from web runtime.");
    return "dt.missing.app.version";
  }
}

// packages/platform/app-environment/src/lib/get-environment-id.ts
function getEnvironmentId() {
  const runtimeFunc = getGlobalWithDtRuntime().dtRuntime?.appEnvironment?.getEnvironmentId;
  if (typeof runtimeFunc === "function") {
    return runtimeFunc();
  } else {
    console.warn("Missing getEnvironmentId function from web runtime.");
    return "dt.missing.environment.id";
  }
}

// packages/platform/app-environment/src/lib/get-environment-url.ts
function getEnvironmentUrl() {
  const runtimeFunc = getGlobalWithDtRuntime().dtRuntime?.appEnvironment?.getEnvironmentUrl;
  if (typeof runtimeFunc === "function") {
    return runtimeFunc();
  } else {
    console.warn("Missing getEnvironmentUrl function from web runtime.");
    return "https://dynatrace.com/";
  }
}

// packages/platform/app-environment/src/lib/get-current-user-details.ts
function getCurrentUserDetails() {
  const runtimeFunc = getGlobalWithDtRuntime().dtRuntime?.appEnvironment?.getCurrentUserDetails;
  if (typeof runtimeFunc === "function") {
    return runtimeFunc();
  } else {
    console.warn("Missing getCurrentUserDetails function from web runtime.");
    return {
      id: "dt.missing.user.id",
      name: "dt.missing.user.name",
      email: "dt.missing.user.email"
    };
  }
}
export {
  getAppId,
  getAppName,
  getAppVersion,
  getCurrentUserDetails,
  getEnvironmentId,
  getEnvironmentUrl
};
