// @dynatrace/strato-components/filters/filter/FilterBar.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var FilterBar_exports = {};
__export(FilterBar_exports, {
  FilterBar: () => FilterBar
});
module.exports = __toCommonJS(FilterBar_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_react_is = require("react-is");
var import_user_preferences = require("@dynatrace-sdk/user-preferences");
var import_FilterBarContext = require("./contexts/FilterBarContext.js");
var import_Item = require("./Item.js");
var import_MoreMenu = require("./MoreMenu.js");
var import_ResetButton = require("./ResetButton.js");
var import_filter_state_reducer = require("./state/filter-state-reducer.js");
var import_filter_state_types = require("./types/filter-state.types.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_is_slot = require("../../core/utils/is-slot.js");
var import_Flex = require("../../layouts/flex/Flex.js");
const COMPONENT_NAME = "FilterBar";
function getItemSlotsAndDefaultPinnedFilters(children, defaultPinnedState) {
  const slots = {
    items: [],
    more: [],
    reset: null
  };
  const defaultPinnedFilters = [];
  for (const child of import_react.Children.toArray(children)) {
    if ((0, import_is_slot.isSlot)(child, import_Item.Item)) {
      const pinnedState = defaultPinnedState[child.props.name] ?? "pinned";
      switch (pinnedState) {
        case "optional":
          slots.more.push(child);
          break;
        case "pinned-optional":
          defaultPinnedFilters.push(child.props.name);
          slots.more.push(child);
          break;
        default:
          slots.items.push(child);
      }
    } else if ((0, import_react_is.isElement)(child) && child.type === import_ResetButton.ResetButton) {
      slots.reset = child;
    } else if (child !== null) {
      console.warn("The used child is not supported by the FilterBar.");
    }
  }
  return { slots, defaultPinnedFilters };
}
const _FilterBar = /* @__PURE__ */ (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const {
      children,
      id,
      onFilterChange,
      onPinnedStateChange,
      defaultPinnedState,
      pinnedState,
      showLabels = true,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
    const language = (0, import_user_preferences.getLanguage)();
    const [lastAddedItem, setLastAddedItem] = (0, import_react.useState)(null);
    const [isInitialized, setIsInitialized] = (0, import_react.useState)(false);
    const callback = (0, import_react.useCallback)(
      (state2, action) => (
        // eslint-disable-next-line @typescript-eslint/prefer-ts-expect-error, @typescript-eslint/ban-ts-comment
        // @ts-ignore Strict mode transition - needs to be investigated
        (0, import_filter_state_reducer.reducer)(state2, action, !!pinnedState, onPinnedStateChange)
      ),
      // pinnedState is excluded since if it would change the component would switch from controlled to uncontrolled.
      // eslint-disable-next-line react-hooks/exhaustive-deps
      [onFilterChange, onPinnedStateChange]
    );
    const [state, dispatch] = (0, import_react.useReducer)(callback, {
      filters: void 0,
      itemsPinnedState: defaultPinnedState ?? pinnedState
    });
    (0, import_react.useEffect)(() => {
      if (pinnedState) {
        dispatch({
          type: import_filter_state_types.FilterBarActions.CHANGEPINNEDSTATE,
          payload: {
            newPinnedState: pinnedState
          }
        });
      }
    }, [pinnedState]);
    (0, import_react.useEffect)(() => {
      if (state.filters === void 0) {
        return;
      }
      if (!isInitialized) {
        setIsInitialized(true);
      } else {
        onFilterChange(state.filters);
      }
    }, [state.filters]);
    const { slots } = (0, import_react.useMemo)(
      () => getItemSlotsAndDefaultPinnedFilters(
        children,
        defaultPinnedState ?? pinnedState ?? {}
      ),
      /**
       * Disabled for defaultPinnedState, since this should only be considered when rendered initially.
       */
      // eslint-disable-next-line react-hooks/exhaustive-deps
      [children, pinnedState]
    );
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_FilterBarContext.FilterBarContext.Provider,
      {
        value: {
          state,
          showLabels,
          dispatch,
          lastAddedItem,
          setLastAddedItem,
          hasControlledPinnedState: !!pinnedState
        },
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
          import_Flex.Flex,
          {
            flexWrap: "wrap",
            id,
            ref: forwardedRef,
            "data-testid": dataTestId,
            "data-dtrum-mask": dataDtrumMask,
            "data-dtrum-allow": dataDtrumAllow,
            alignItems: "end",
            className: consumerClassName,
            style: consumerStyle,
            ...behavioralTrackingProps,
            "data-dt-component": COMPONENT_NAME,
            children: [
              slots.items,
              slots.more?.length > 0 && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_intl.IntlProvider, { locale: language, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_MoreMenu.MoreMenu, { children: slots.more }) }),
              slots.reset
            ]
          }
        )
      }
    );
  }
);
_FilterBar.displayName = COMPONENT_NAME;
const FilterBar = Object.assign(_FilterBar, {
  Item: import_Item.Item,
  ResetButton: import_ResetButton.ResetButton
});
