// @dynatrace/strato-components/filters/filter/Item.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Item_exports = {};
__export(Item_exports, {
  Item: () => Item
});
module.exports = __toCommonJS(Item_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react2 = require("react");
var import_useFilterBarContext = require("./hooks/useFilterBarContext.js");
var import_filter_state_types = require("./types/filter-state.types.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useControllableState = require("../../core/hooks/useControllableState.js");
var import_is_node_environment = require("../../core/utils/is-node-environment.js");
var import_merge_props = require("../../core/utils/merge-props.js");
var import_FormField = require("../../forms/forms-core/components/form-field/FormField.js");
var import_Label = require("../../forms/forms-core/components/label/Label.js");
var import_shared_types = require("../../forms/shared-types.js");
const InternalItem = (props) => {
  const {
    name,
    label,
    children: child,
    showLabel: propShowLabel,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const { defaultValue, value, onChange, ...propsWithoutContractProperties } = child.props;
  const [filterValue, setFilterValue] = (0, import_useControllableState.useControllableState)({
    defaultValue,
    value,
    onChange
  });
  const { dispatch, showLabels, lastAddedItem, state } = (0, import_useFilterBarContext.useFilterBarContext)();
  const showLabel = propShowLabel ?? showLabels;
  const isStringLabel = typeof label === "string";
  const filterItemRef = (0, import_react2.useRef)(
    null
  );
  if (!isStringLabel && !showLabel && (0, import_is_node_environment.isDev)()) {
    console.warn(
      "FilterBar.Item: the showLabel prop was set to false but the provided label can't be used as aria-label because it is not a string. Please provide a string-based label or set showLabel to true for this FilterBar.Item."
    );
  }
  (0, import_react2.useEffect)(() => {
    dispatch({
      type: import_filter_state_types.FilterBarActions.INITIALIZE,
      payload: {
        name,
        value: filterValue
      }
    });
    setTimeout(() => {
      if (lastAddedItem === name) {
        filterItemRef?.current?.focus?.();
        if (filterItemRef?.current && (0, import_shared_types.isFormControlWithOverlayRef)(filterItemRef?.current)) {
          filterItemRef?.current?.open();
        }
      }
    });
    return () => {
      dispatch({
        type: import_filter_state_types.FilterBarActions.REMOVE,
        payload: {
          name
        }
      });
    };
  }, [name]);
  (0, import_react2.useEffect)(() => {
    if (state.filters?.[name] && filterValue !== state.filters?.[name].value) {
      dispatch({
        type: import_filter_state_types.FilterBarActions.CHANGE,
        payload: {
          name,
          value: filterValue
        }
      });
    }
  }, [filterValue, state.filters?.[name]]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_FormField.FormField,
    {
      "data-testid": dataTestId,
      id: propsWithoutContractProperties.id,
      className: consumerClassName,
      style: consumerStyle,
      ...behavioralTrackingProps,
      children: [
        showLabel && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Label.Label, { children: label }),
        (0, import_react2.cloneElement)(child, {
          ...(0, import_merge_props.mergeProps)(propsWithoutContractProperties, {
            value: filterValue,
            onChange: setFilterValue,
            "aria-label": isStringLabel && !showLabel ? label : void 0,
            ref: filterItemRef
          })
        })
      ]
    }
  );
};
const Item = (props) => {
  return /* @__PURE__ */ (0, import_react.createElement)(InternalItem, { ...props, key: props.name });
};
Item["displayName"] = "FilterBar.Item";
