// @dynatrace/strato-components/filters/filter/MoreMenu.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MoreMenu_exports = {};
__export(MoreMenu_exports, {
  MoreMenu: () => MoreMenu
});
module.exports = __toCommonJS(MoreMenu_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_useFilterBarContext = require("./hooks/useFilterBarContext.js");
var import_Item = require("./Item.js");
var import_filter_state_types = require("./types/filter-state.types.js");
var import_getTextFromNode = require("./utils/getTextFromNode.js");
var import_Button = require("../../buttons/button/Button.js");
var import_useId = require("../../core/hooks/useId.js");
var import_useFocusRing = require("../../core/styles/useFocusRing.js");
var import_is_slot = require("../../core/utils/is-slot.js");
var import_Select = require("../../forms/select/Select.js");
var import_Flex = require("../../layouts/flex/Flex.js");
const MoreMenu = ({ children }) => {
  const intl = (0, import_react_intl.useIntl)();
  const label = intl.formatMessage({
    defaultMessage: "Add filter",
    id: "+BUznwQEJTiILiAa",
    description: "Message to display by default for adding more Filters to the FilterBar."
  });
  const { focusClassName, focusProps } = (0, import_useFocusRing.useFocusRing)({
    color: "neutral",
    focusWithin: true,
    ignoreModality: true
  });
  const { dispatch, setLastAddedItem, state } = (0, import_useFilterBarContext.useFilterBarContext)(), { itemsPinnedState } = state;
  const selectRef = (0, import_react.useRef)(null);
  const allPinnableFilters = (0, import_react.useMemo)(() => {
    const temp = {};
    for (const child of import_react.Children.toArray(children)) {
      if ((0, import_is_slot.isSlot)(child, import_Item.Item)) {
        const uniqueName = child.props.name;
        temp[uniqueName] = {
          elem: child,
          label: (0, import_getTextFromNode.getTextFromNode)(child.props.label, intl, uniqueName)
        };
      }
    }
    return temp;
  }, [children, intl]);
  const optionalPinnedItems = (0, import_react.useMemo)(
    () => Object.entries(itemsPinnedState || {}).reduce((keys, [key, value]) => {
      if (value === "pinned-optional") {
        keys.push(key);
      }
      return keys;
    }, []),
    [itemsPinnedState]
  );
  const handleSelectChange = (newValues) => {
    setLastAddedItem(null);
    const union = /* @__PURE__ */ new Set([
      ...newValues ?? [],
      ...optionalPinnedItems || []
    ]);
    if (!optionalPinnedItems || newValues && optionalPinnedItems.length < newValues.length) {
      const filtersToPin = [...union].filter(
        (key) => !optionalPinnedItems?.includes(key)
      );
      selectRef?.current?.close();
      filtersToPin?.length && setLastAddedItem(filtersToPin[0]?.toString());
      dispatch({
        type: import_filter_state_types.FilterBarActions.PIN,
        payload: {
          keys: filtersToPin
        }
      });
      return;
    }
    if (!newValues) {
      dispatch({
        type: import_filter_state_types.FilterBarActions.UNPIN,
        payload: {
          keys: optionalPinnedItems ?? []
        }
      });
    } else if (optionalPinnedItems && optionalPinnedItems.length > newValues.length) {
      const keys = [...union].filter((key) => !newValues?.includes(key));
      dispatch({
        type: import_filter_state_types.FilterBarActions.UNPIN,
        payload: { keys }
      });
    }
  };
  const labelId = (0, import_useId.useId)();
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    optionalPinnedItems.map(
      (key) => allPinnableFilters[key] && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Flex.Flex, { gap: 2, alignItems: "flex-end", children: allPinnableFilters[key].elem }, key)
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { style: { maxWidth: "100%" }, ...focusProps, children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      import_Select.Select,
      {
        value: optionalPinnedItems,
        onChange: handleSelectChange,
        id: "user-select",
        multiple: true,
        clearable: true,
        ref: selectRef,
        "data-dt-name": "AddFilter",
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Select.Select.CustomTrigger, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Button.Button, { as: "div", id: labelId, className: focusClassName, children: [
            label,
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Suffix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ChevronDownSmallIcon, {}) })
          ] }) }),
          /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Select.Select.Content, { width: "fit-content", children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Select.Select.Filter, {}),
            Object.keys(allPinnableFilters).map((key) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Select.Select.Option, { value: key, children: allPinnableFilters[key].label }, key))
          ] })
        ]
      }
    ) })
  ] });
};
MoreMenu.displayName = "FilterBar.MoreMenu";
