// @dynatrace/strato-components/filters/filter/ResetButton.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ResetButton_exports = {};
__export(ResetButton_exports, {
  ResetButton: () => ResetButton
});
module.exports = __toCommonJS(ResetButton_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_Button = require("../../buttons/button/Button.js");
var import_useId = require("../../core/hooks/useId.js");
var import_Tooltip = require("../../overlays/tooltip/Tooltip.js");
const ResetButton = /* @__PURE__ */ (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const {
      onClick,
      "aria-labelledby": ariaLabelledByProp,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      ...remainingProps
    } = props;
    const intl = (0, import_react_intl.useIntl)();
    const resetText = intl.formatMessage({
      defaultMessage: "Reset filters",
      id: "21mG229oVCMPPsYY",
      description: "Shown for the reset button of the FilterBar."
    });
    const tooltipId = (0, import_useId.useId)();
    const ariaLabelledBy = (0, import_react.useMemo)(
      () => tooltipId || ariaLabelledByProp ? `${tooltipId} ${ariaLabelledByProp ?? ""}` : void 0,
      [ariaLabelledByProp, tooltipId]
    );
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Tooltip.Tooltip, { id: tooltipId, text: resetText, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Button.Button,
      {
        "data-testid": dataTestId,
        ref: forwardedRef,
        variant: "default",
        onClick,
        "aria-label": resetText,
        "aria-labelledby": ariaLabelledBy,
        className: consumerClassName,
        style: consumerStyle,
        ...remainingProps,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ResetIcon, {}) })
      }
    ) });
  }
);
ResetButton.displayName = "FilterBar.ResetButton";
