// @dynatrace/strato-components/charts/histogram/utils/histogram-configuration.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var histogram_configuration_exports = {};
__export(histogram_configuration_exports, {
  createHistogramConfiguration: () => createHistogramConfiguration,
  deserializeConfig: () => deserializeConfig,
  getColorPaletteConfiguration: () => getColorPaletteConfiguration,
  getCurrentZoomConfiguration: () => getCurrentZoomConfiguration,
  getInfiniteZoomConfiguration: () => getInfiniteZoomConfiguration,
  getInitialZoomConfiguration: () => getInitialZoomConfiguration,
  getLegendConfiguration: () => getLegendConfiguration,
  getThresholdsConfiguration: () => getThresholdsConfiguration,
  getTooltipConfiguration: () => getTooltipConfiguration,
  getXAxisConfiguration: () => getXAxisConfiguration,
  getYAxisConfiguration: () => getYAxisConfiguration,
  isValidHistogramTooltipVariant: () => isValidHistogramTooltipVariant,
  normalizeConfig: () => normalizeConfig
});
module.exports = __toCommonJS(histogram_configuration_exports);
var import_lodash_es = require("lodash");
var import_get_toolbar_configuration = require("./get-toolbar-configuration.js");
var import_constants = require("../../core/constants.js");
var import_ColorRule = require("../../core/slots/ColorRule.js");
var import_color_palette = require("../../core/types/color-palette.js");
var import_getValidColor = require("../../core/utils/getValidColor.js");
var import_interactions = require("../../core/utils/interactions.js");
var import_is_valid_color_palette = require("../../core/utils/is-valid-color-palette.js");
var import_is_valid_legend_position = require("../../core/utils/is-valid-legend-position.js");
var import_defaults = require("../constants/defaults.js");
var import_histogram = require("../types/histogram.js");
function getColorPaletteConfiguration(contextColorPalette, propsColorPalette) {
  const colorPalette = propsColorPalette ?? contextColorPalette ?? import_defaults.defaultColorPalette;
  return (0, import_color_palette.isCustomColorPalette)(colorPalette) || (0, import_is_valid_color_palette.isValidColorPalette)(colorPalette) ? colorPalette : import_defaults.defaultHistogramChartConfig.colorPalette ?? import_defaults.defaultColorPalette;
}
function getLegendConfiguration(contextLegend, slotsLegend) {
  const legendConfig = (0, import_lodash_es.merge)(
    {},
    import_defaults.defaultLegend,
    contextLegend,
    slotsLegend
  );
  legendConfig.position = (0, import_is_valid_legend_position.isValidLegendPosition)(legendConfig.position) ? legendConfig.position : import_defaults.defaultLegend.position;
  return legendConfig;
}
const isValidHistogramTooltipVariant = (tooltipVariant) => import_histogram.histogramTooltipVariantOptions.includes(
  tooltipVariant
);
function getTooltipConfiguration(contextTooltip, slotsTooltip) {
  const tooltip = (0, import_lodash_es.merge)(
    {},
    { variant: import_defaults.defaultTooltipVariant },
    contextTooltip,
    slotsTooltip
  );
  tooltip.variant = isValidHistogramTooltipVariant(tooltip.variant) ? tooltip.variant : import_defaults.defaultTooltipVariant;
  return tooltip;
}
const sanitizeYAxisConfig = (yAxes) => {
  const reversedYAxes = yAxes.reverse();
  const leftAxis = reversedYAxes.find(
    (axis) => axis?.position === "left" || axis?.position === void 0
  );
  const rightAxis = reversedYAxes.find((axis) => axis?.position === "right");
  const sanitizedYAxes = {};
  if (leftAxis) {
    sanitizedYAxes.left = leftAxis;
  }
  if (rightAxis) {
    sanitizedYAxes.right = rightAxis;
  }
  return sanitizedYAxes;
};
function getYAxisConfiguration(slotsYAxis, contextYAxis) {
  const contextConfig = contextYAxis === void 0 ? {} : sanitizeYAxisConfig(contextYAxis);
  const slotsConfig = slotsYAxis === void 0 ? {} : sanitizeYAxisConfig(slotsYAxis);
  const { left, right } = (0, import_lodash_es.merge)({}, contextConfig, slotsConfig);
  if (left === void 0 && right === void 0) {
    return import_defaults.defaultYAxis;
  }
  const result = {};
  if (left) {
    result.left = (0, import_lodash_es.merge)({}, import_defaults.defaultYAxis.left, left);
  }
  if (right) {
    const defaultRight = { ...import_defaults.defaultYAxis.left, position: "right" };
    result.right = (0, import_lodash_es.merge)({}, defaultRight, right);
  }
  return result;
}
function getXAxisConfiguration(contextXAxis, slotsXAxis) {
  return (0, import_lodash_es.merge)({}, import_defaults.defaultXAxis, contextXAxis, slotsXAxis);
}
function getThresholdsConfiguration(contextThresholds = [], slotsThresholds = []) {
  const allThresholds = contextThresholds.concat(slotsThresholds).map((threshold) => ({
    ...threshold,
    id: crypto.randomUUID(),
    position: threshold.position ? threshold.position : "left",
    color: (0, import_getValidColor.isValidCSSColor)(threshold.color) ? threshold.color : import_defaults.DEFAULT_SCALE_FALLBACK_COLOR
  }));
  if (!allThresholds.length) {
    return void 0;
  }
  const thresholdConfig = {};
  thresholdConfig.leftThreshold = allThresholds.filter(
    (threshold) => threshold.position === "left"
  );
  thresholdConfig.rightThreshold = allThresholds.filter(
    (threshold) => threshold.position === "right"
  );
  return thresholdConfig;
}
const getValueRepresentation = (contextValueRepresentation, configValueRepresentation) => contextValueRepresentation ?? configValueRepresentation ?? import_constants.DEFAULT_VALUE_REPRESENTATION;
function getInfiniteZoomConfiguration(contextConfig, propsConfig) {
  return propsConfig ?? contextConfig ?? import_constants.DEFAULT_INFINITE_ZOOM;
}
function getInitialZoomConfiguration(contextConfig, propsConfig) {
  return propsConfig ?? contextConfig;
}
function getCurrentZoomConfiguration(contextConfig, propsConfig) {
  return propsConfig ?? contextConfig;
}
function createHistogramConfiguration(slots, contextConfig, histogramChartConfigProps) {
  const colorPalette = getColorPaletteConfiguration(
    contextConfig.colorPalette,
    histogramChartConfigProps?.colorPalette
  );
  const legend = getLegendConfiguration(contextConfig.legend, slots.legend);
  const tooltip = getTooltipConfiguration(contextConfig.tooltip, slots.tooltip);
  const yAxis = getYAxisConfiguration(slots.yAxis, contextConfig.yAxis);
  const xAxis = getXAxisConfiguration(contextConfig.xAxis, slots.xAxis);
  const truncationMode = histogramChartConfigProps?.truncationMode ?? contextConfig?.truncationMode ?? import_constants.DEFAULT_TRUNCATION_MODE;
  const toolbar = (0, import_get_toolbar_configuration.getToolbarConfiguration)(
    contextConfig.toolbar,
    slots.toolbar,
    slots.downloadCSV,
    contextConfig.downloadCSV
  );
  const propsInteractions = {
    onZoomChange: histogramChartConfigProps.onZoomChange
  };
  const hasContextInteractions = contextConfig && (contextConfig.pan || contextConfig.zoom);
  const chartConfigInteractions = hasContextInteractions ? {
    pan: { enabled: contextConfig.pan?.disabled !== true },
    zoom: { enabled: contextConfig.zoom?.disabled !== true },
    zoomX: { enabled: contextConfig.zoom?.disabled !== true }
  } : {};
  const interactions = (0, import_interactions.getInteractions)(
    contextConfig.interactions,
    slots.interactions ?? {},
    propsInteractions,
    chartConfigInteractions
  );
  const configThresholds = contextConfig.thresholdIndicators || [];
  const thresholds = getThresholdsConfiguration(
    configThresholds,
    slots.thresholds
  );
  const valueRepresentation = getValueRepresentation(
    contextConfig.valueRepresentation,
    histogramChartConfigProps.valueRepresentation
  );
  const initialZoom = getInitialZoomConfiguration(
    contextConfig.initialZoom,
    histogramChartConfigProps.initialZoom
  );
  const currentZoom = getCurrentZoomConfiguration(
    contextConfig.currentZoom,
    histogramChartConfigProps.currentZoom
  );
  const infiniteZoom = getInfiniteZoomConfiguration(
    contextConfig.infiniteZoom,
    histogramChartConfigProps.infiniteZoom
  );
  const colorRules = (contextConfig.colorRules || []).filter(import_ColorRule.isValidColorRule).concat(slots.colorRules || []);
  return {
    valueRepresentation,
    colorPalette,
    legend,
    tooltip,
    yAxis,
    xAxis,
    truncationMode,
    toolbar,
    annotations: slots.annotations,
    interactions,
    thresholds,
    initialZoom,
    currentZoom,
    infiniteZoom,
    select: slots.select,
    colorRules
  };
}
function deserializeConfig(serializedConfig) {
  try {
    const parsedData = JSON.parse(serializedConfig);
    parsedData.interactions = {
      ...parsedData.interactions,
      zoom: { enabled: parsedData.zoom?.disabled !== true },
      zoomX: { enabled: parsedData.zoom?.disabled !== true }
    };
    return (0, import_lodash_es.merge)({}, import_defaults.defaultHistogramChartConfig, parsedData);
  } catch {
    return import_defaults.defaultHistogramChartConfig;
  }
}
function parseYAxisConfig(yAxis) {
  const { left, right } = yAxis;
  const yAxisConfig = [];
  if (left) {
    yAxisConfig.push(left);
  }
  if (right) {
    yAxisConfig.push(right);
  }
  return yAxisConfig;
}
function parseThresholdsConfig(thresholds) {
  const { leftThreshold, rightThreshold } = thresholds;
  const thresholdsConfig = [];
  if (leftThreshold) {
    thresholdsConfig.push(...leftThreshold);
  }
  if (rightThreshold) {
    thresholdsConfig.push(...rightThreshold);
  }
  return thresholdsConfig;
}
function normalizeConfig(config) {
  const { thresholds, ...rest } = config;
  const thresholdIndicators = thresholds ? parseThresholdsConfig(thresholds) : void 0;
  const { intents, ...toolbar } = config.toolbar ?? {};
  return {
    ...rest,
    toolbar,
    yAxis: parseYAxisConfig(config.yAxis),
    thresholdIndicators
  };
}
