// @dynatrace/strato-components/charts/histogram/utils/iterate-slots.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var iterate_slots_exports = {};
__export(iterate_slots_exports, {
  iterateSlots: () => iterateSlots
});
module.exports = __toCommonJS(iterate_slots_exports);
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_react_is = require("react-is");
var import_Intent = require("../../../core/slots/Intent/Intent.js");
var import_is_slot = require("../../../core/utils/is-slot.js");
var import_populate_missing_marker_props = require("../../core/components/annotations/utils/populate-missing-marker-props.js");
var import_populate_missing_track_props = require("../../core/components/annotations/utils/populate-missing-track-props.js");
var import_ChartInteractions = require("../../core/slots/chart-interactions/ChartInteractions.js");
var import_ColorRule = require("../../core/slots/ColorRule.js");
var import_ChartToolbar = require("../../core/slots/toolbar/ChartToolbar.js");
var import_DownloadData = require("../../core/slots/toolbar/DownloadData.js");
var import_interactions2 = require("../../core/utils/interactions.js");
var import_iterate_chart_interactions_slots = require("../../core/utils/iterate-chart-interactions-slots.js");
var import_Annotations = require("../slots/annotations/Annotations.js");
var import_Marker = require("../slots/annotations/Marker.js");
var import_Tooltip = require("../slots/annotations/Tooltip.js");
var import_Track = require("../slots/annotations/Track.js");
var import_DownloadCSV = require("../slots/DownloadCSV.js");
var import_Legend = require("../slots/Legend.js");
var import_Pan = require("../slots/Pan.js");
var import_Select = require("../slots/Select.js");
var import_ThresholdIndicator = require("../slots/ThresholdIndicator.js");
var import_Toolbar = require("../slots/Toolbar.js");
var import_Tooltip2 = require("../slots/Tooltip.js");
var import_XAxis = require("../slots/XAxis.js");
var import_YAxis = require("../slots/YAxis.js");
var import_Zoom = require("../slots/Zoom.js");
const slotInteractionsDefaultValues = {
  zoom: { enabled: true },
  zoomX: { enabled: true },
  pan: { enabled: true }
};
const chartInteractionsDefaultValues = {
  zoom: { enabled: false },
  zoomX: { enabled: false },
  pan: { enabled: false }
};
const getInteractionsConfig = (chartConfig, slotsConfig) => {
  const chartApiValues = (0, import_interactions2.getConfigPlusDefaultOrEmpty)(
    chartInteractionsDefaultValues,
    chartConfig
  );
  const slotsApiValues = (0, import_interactions2.getConfigPlusDefaultOrEmpty)(
    slotInteractionsDefaultValues,
    slotsConfig
  );
  return (0, import_lodash_es.merge)({}, chartApiValues, slotsApiValues);
};
const disabledOnlyWhenSpecified = ({ disabled }) => disabled;
const iterateSlots = (children) => {
  let slots = {
    legend: {},
    xAxis: {},
    yAxis: [],
    tooltip: {},
    toolbar: {},
    interactions: {},
    thresholds: [],
    colorRules: []
  };
  let slotToolbar = {};
  let chartToolbar = {};
  let chartInteractions = {};
  let slotInteractions = {};
  let intents;
  import_react.Children.forEach(children, (child) => {
    if ((0, import_is_slot.isSlot)(child, import_ChartInteractions.ChartInteractions)) {
      const { children: children2, ...previousInteractions } = child.props;
      const interactions = (0, import_iterate_chart_interactions_slots.iterateChartInteractionsSlots)(children2);
      chartInteractions = {
        ...previousInteractions,
        ...interactions
      };
    } else if ((0, import_is_slot.isSlot)(child, import_XAxis.XAxis)) {
      slots = {
        ...slots,
        xAxis: child.props
      };
    } else if ((0, import_is_slot.isSlot)(child, import_YAxis.YAxis)) {
      slots.yAxis.push(child.props);
    } else if ((0, import_is_slot.isSlot)(child, import_Legend.Legend)) {
      slots = {
        ...slots,
        legend: child.props
      };
    } else if ((0, import_is_slot.isSlot)(child, import_Tooltip2.Tooltip)) {
      slots = {
        ...slots,
        tooltip: child.props
      };
    } else if ((0, import_is_slot.isSlot)(child, import_ChartToolbar.ChartToolbar)) {
      const { children: children2, ...toolbar } = child.props;
      const childrenToolbar = iterateChartToolbarSlots(children2);
      chartToolbar = {
        enabled: true,
        ...toolbar,
        ...childrenToolbar
      };
    } else if ((0, import_is_slot.isSlot)(child, import_Toolbar.Toolbar)) {
      slotToolbar = {
        enabled: child.props.hidden !== true,
        ...slotToolbar,
        ...child.props
      };
    } else if ((0, import_is_slot.isSlot)(child, import_Zoom.Zoom)) {
      const enabled = !disabledOnlyWhenSpecified(child.props);
      slotInteractions = {
        ...slotInteractions,
        zoom: { enabled },
        zoomX: { enabled }
      };
    } else if ((0, import_is_slot.isSlot)(child, import_Pan.Pan)) {
      const enabled = !disabledOnlyWhenSpecified(child.props);
      slotInteractions = { ...slotInteractions, pan: { enabled } };
    } else if ((0, import_is_slot.isSlot)(child, import_DownloadCSV.HistogramChartDownloadCSVSlot)) {
      slotToolbar = { ...slotToolbar, downloadData: { enabled: true } };
    } else if ((0, import_is_slot.isSlot)(child, import_Intent.Intent)) {
      intents = intents ? [...intents, child.props] : [child.props];
    } else if ((0, import_is_slot.isSlot)(child, import_Annotations.Annotations)) {
      const { children: children2, ...annotationsProps } = child.props;
      const { tracks, tooltip } = iterateTrackSlots(children2);
      const annotations = {
        ...annotationsProps,
        tracks,
        tooltip
      };
      slots = {
        ...slots,
        // eslint-disable-next-line @typescript-eslint/prefer-ts-expect-error, @typescript-eslint/ban-ts-comment
        // @ts-ignore Strict mode transition - needs to be investigated
        annotations
      };
    } else if ((0, import_is_slot.isSlot)(child, import_ThresholdIndicator.ThresholdIndicator)) {
      slots.thresholds.push(child.props);
    } else if ((0, import_is_slot.isSlot)(child, import_Select.Select)) {
      slots = {
        ...slots,
        select: child.props
      };
    } else if ((0, import_is_slot.isSlot)(child, import_ColorRule.ColorRule)) {
      (0, import_ColorRule.isValidColorRule)(child.props) && slots.colorRules.push(child.props);
    }
  });
  const hasInteractionConfig = Object.keys(chartInteractions).length > 0 || Object.keys(slotInteractions).length > 0;
  slots.interactions = hasInteractionConfig ? getInteractionsConfig(chartInteractions, slotInteractions) : {};
  const hasToolbarConfig = !(0, import_lodash_es.isEmpty)(slotToolbar) || !(0, import_lodash_es.isEmpty)(chartToolbar) || intents;
  if (hasToolbarConfig) {
    slots.toolbar = {
      ...chartToolbar,
      ...slotToolbar,
      enabled: slotToolbar.enabled ?? chartToolbar.enabled ?? true,
      intents
    };
  }
  return slots;
};
const iterateChartToolbarSlots = (children) => {
  let toolbar = {};
  import_react.Children.forEach(children, (childToolbar) => {
    if ((0, import_is_slot.isSlot)(childToolbar, import_DownloadData.DownloadData)) {
      toolbar = {
        ...toolbar,
        downloadData: { enabled: true }
      };
    }
  });
  return toolbar;
};
const iterateTrackSlots = (children) => {
  const tracks = [];
  let tooltip = {
    template: void 0,
    hidden: false
  };
  import_react.Children.forEach(children, (child) => {
    if ((0, import_is_slot.isSlot)(child, import_Track.Track)) {
      const { children: children2, ...trackProps } = child.props;
      const currentTrack = {
        ...trackProps,
        markers: []
      };
      import_react.Children.forEach(children2, (child2) => {
        if ((0, import_react_is.isElement)(child2)) {
          if ((0, import_is_slot.isSlot)(child2, import_Marker.Marker)) {
            const marker = (0, import_populate_missing_marker_props.populateMissingMarkerProps)(child2.props, trackProps);
            currentTrack.markers.push(marker);
          }
        }
      });
      const track = (0, import_populate_missing_track_props.populateMissingTrackProps)(currentTrack);
      tracks.push(track);
    } else if ((0, import_is_slot.isSlot)(child, import_Tooltip.AnnotationsTooltip)) {
      const { children: children2, hidden } = child.props;
      tooltip = {
        template: children2,
        hidden: hidden ?? false
      };
    }
  });
  return { tracks, tooltip };
};
