// @dynatrace/strato-components/esm/buttons/run-query-button/RunQueryButton.js
import { jsx, jsxs } from "react/jsx-runtime";
import { forwardRef } from "react";
import { FormattedMessage, useIntl } from "react-intl";
import { CriticalIcon, PlayIcon } from "@dynatrace/strato-icons";
import { ProgressCircle } from "../../content/progress/ProgressCircle.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { mergeProps } from "../../core/utils/merge-props.js";
import { Button } from "../button/Button.js";
const COMPONENT_NAME = "RunQueryButton";
const queryStateOptions = [
  "idle",
  "loading",
  "success",
  "error"
];
function getVariant(queryState) {
  switch (queryState) {
    case "idle":
      return "accent";
    case "loading":
    case "success":
    case "error":
      return "emphasized";
  }
}
function getColor(queryState) {
  switch (queryState) {
    case "idle":
      return "primary";
    case "loading":
    case "success":
      return "neutral";
    case "error":
      return "critical";
  }
}
function PrefixIcon({
  type,
  ariaLabel
}) {
  switch (type) {
    case "idle":
    case "success":
      return /* @__PURE__ */ jsx(PlayIcon, {});
    case "loading":
      return /* @__PURE__ */ jsx(ProgressCircle, { size: "small", "aria-label": ariaLabel });
    case "error":
      return /* @__PURE__ */ jsx(CriticalIcon, {});
  }
}
function ButtonMessage(props) {
  const { type, cancelable } = props;
  switch (type) {
    case "idle":
      return /* @__PURE__ */ jsx(
        FormattedMessage,
        {
          description: "Label for button used to run a query",
          defaultMessage: "Run query",
          id: "CwGi8xbTtV+fdsu5"
        }
      );
    case "success":
      return /* @__PURE__ */ jsx(
        FormattedMessage,
        {
          description: "Label for button used to rerun a query after it was successful",
          defaultMessage: "Run query",
          id: "J9M12SSZNn2dKokI"
        }
      );
    case "loading":
      return cancelable ? /* @__PURE__ */ jsx(
        FormattedMessage,
        {
          description: "Call to action for cancelable query button",
          defaultMessage: "Cancel query",
          id: "Za5bQfBDnyyur0mw"
        }
      ) : /* @__PURE__ */ jsx(
        FormattedMessage,
        {
          description: "Label for button indicating that a query is loading",
          defaultMessage: "Running",
          id: "hi3IodOo0ygpihhY"
        }
      );
    case "error":
      return /* @__PURE__ */ jsx(
        FormattedMessage,
        {
          description: "Label for button indicating that an error has occurred while running the query. Click to retry.",
          defaultMessage: "Error (click to retry)",
          id: "gg76gjIy5leuPbMT"
        }
      );
  }
}
const RunQueryButton = /* @__PURE__ */ forwardRef((props, ref) => {
  const {
    queryState = "idle",
    disabled = false,
    onClick,
    cancelable = true,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
  const allowsClickHandler = queryState !== "loading" || queryState === "loading" && cancelable;
  const intl = useIntl();
  const ariaLabel = intl.formatMessage({
    description: "Label for button indicating that a query is loading",
    defaultMessage: "Running",
    id: "hi3IodOo0ygpihhY"
  });
  return /* @__PURE__ */ jsxs(
    Button,
    {
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      onClick: allowsClickHandler ? onClick : void 0,
      color: getColor(queryState),
      variant: getVariant(queryState),
      ref,
      disabled,
      "aria-live": "polite",
      "aria-atomic": "true",
      className: consumerClassName,
      style: consumerStyle,
      ...mergeProps(restProps, behavioralTrackingProps),
      "data-dt-component": COMPONENT_NAME,
      children: [
        /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(PrefixIcon, { type: queryState, ariaLabel }) }),
        /* @__PURE__ */ jsx(ButtonMessage, { type: queryState, cancelable })
      ]
    }
  );
});
RunQueryButton.displayName = COMPONENT_NAME;
export {
  RunQueryButton,
  queryStateOptions
};
