// @dynatrace/strato-components/esm/navigation/breadcrumbs/Breadcrumbs.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import {
  Children,
  cloneElement,
  forwardRef,
  useMemo,
  useRef
} from "react";
import { useIntl } from "react-intl";
import Colors from "@dynatrace/strato-design-tokens/colors";
import {
  breadcrumbsContainerCSS,
  breadcrumbsItemInvisibleCSS
} from "./Breadcrumbs.sty.js";
import { isBreadcrumbItem, Item } from "./Item.js";
import { Button } from "../../buttons/button/Button.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useContainerSize } from "../../core/hooks/useContainerSize.js";
import { useMeasureChildren } from "../../core/hooks/useMeasureChildren.js";
import { useMergeRefs } from "../../core/hooks/useMergeRefs.js";
import { Menu } from "../menu/Menu.js";
const COMPONENT_NAME = "Breadcrumbs";
const BreadcrumbsRoot = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const {
    children,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    ...remainingProps
  } = props;
  const _containerRef = useRef(null);
  const { containerRef, measurements, childProps } = useMeasureChildren();
  const { width: containerWidth } = useContainerSize(_containerRef);
  const rootRef = useMergeRefs([forwardedRef, _containerRef, containerRef]);
  const overflowingBreadcrumbs = useRef(/* @__PURE__ */ new Map());
  const intl = useIntl();
  const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
  const slots = useMemo(() => {
    const slots2 = { items: [], menuItems: [] };
    const childrenArray = Children.toArray(children);
    let childWidths = 0;
    for (const value of Object.values(measurements)) {
      childWidths += value?.width || 0;
    }
    if (Object.keys(measurements).length === childProps.length) {
      const childKeys = childProps.map((props2) => props2["data-id"]);
      let childIndex = overflowingBreadcrumbs.current.size ? 2 : 1;
      overflowingBreadcrumbs.current.clear();
      while (childWidths > containerWidth && childIndex < childProps.length) {
        if (childIndex !== childProps.length - 1) {
          const key = childKeys[childIndex];
          if (key) {
            overflowingBreadcrumbs.current.set(key, true);
            childWidths -= measurements[key]?.width || 0;
          }
        } else {
          const key = childKeys[0];
          if (key) {
            overflowingBreadcrumbs.current.set(key, true);
            childWidths -= measurements[key]?.width || 0;
          }
        }
        childIndex++;
      }
    }
    for (const [index, child] of childrenArray.entries()) {
      if (isBreadcrumbItem(child)) {
        const uniqueKey = `nav-items-${index}`;
        const adjustedIndex = overflowingBreadcrumbs.current.size > 0 && index > 0 ? index + 1 : index;
        const isLast = index === childrenArray.length - 1;
        slots2.items.push(
          cloneElement(child, {
            ...child.props,
            key: uniqueKey,
            className: clsx(
              {
                [`${breadcrumbsItemInvisibleCSS}`]: overflowingBreadcrumbs.current.get(
                  childProps[adjustedIndex]?.["data-id"] || ""
                )
              },
              child.props.className
            ),
            isLast,
            isLastVisible: isLast && overflowingBreadcrumbs.current.size === childrenArray.length - 1,
            isCollapsed: false,
            ...childProps[adjustedIndex]
          })
        );
        if (overflowingBreadcrumbs.current.get(
          childProps[adjustedIndex]?.["data-id"] || ""
        )) {
          slots2.menuItems.push(
            cloneElement(/* @__PURE__ */ jsx(Menu.Link, {}), {
              ...child.props,
              key: uniqueKey,
              isCollapsed: true,
              ...childProps[adjustedIndex]
            })
          );
        }
      }
    }
    return slots2;
  }, [children, measurements, containerWidth, childProps]);
  return /* @__PURE__ */ jsx(
    "nav",
    {
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "aria-label": intl.formatMessage({
        defaultMessage: "Breadcrumb",
        id: "x2bmZRg98b0KodlW"
      }),
      className: consumerClassName,
      style: consumerStyle,
      ...restProps,
      ...behavioralTrackingProps,
      "data-dt-component": COMPONENT_NAME,
      children: /* @__PURE__ */ jsxs("ol", { ref: rootRef, className: breadcrumbsContainerCSS, children: [
        slots.items[0],
        overflowingBreadcrumbs.current.size > 0 && /* @__PURE__ */ jsxs(Menu, { children: [
          /* @__PURE__ */ jsx(Menu.Trigger, { children: /* @__PURE__ */ jsx(Item, { as: Button, ...childProps[1], menuTrigger: true, size: "condensed", children: /* @__PURE__ */ jsx(Button.Prefix, { style: { color: Colors.Text.Neutral.Default }, children: "\u2026" }) }) }),
          /* @__PURE__ */ jsx(Menu.Content, { children: slots.menuItems })
        ] }),
        slots.items.slice(1)
      ] })
    }
  );
});
BreadcrumbsRoot.displayName = "Breadcrumbs";
const Breadcrumbs = Object.assign(BreadcrumbsRoot, {
  Item
});
export {
  Breadcrumbs
};
