// @dynatrace/strato-components/esm/navigation/breadcrumbs/Item.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import {
  forwardRef
} from "react";
import { isElement } from "react-is";
import { ChevronRightSmallIcon } from "@dynatrace/strato-icons";
import {
  breadcrumbsItemCSS,
  breadcrumbsListItemCSS,
  breadcrumbsSeperatorIconCSS
} from "./Breadcrumbs.sty.js";
import { focusRingStatic } from "../../core/styles/focusRingStatic.sty.js";
import { Link } from "../../typography/link/Link.js";
function isBreadcrumbItem(node) {
  return node !== void 0 && node !== null && isElement(node) && node.type === Item;
}
const Item = /* @__PURE__ */ forwardRef(
  (props, forwardedRef) => {
    const {
      children,
      disabled = false,
      href,
      isLast,
      isLastVisible,
      isCollapsed,
      className: consumerClassName,
      style: consumerStyle,
      "data-id": dataId,
      "data-testid": dataTestId,
      as = "a",
      ...remainingProps
    } = props;
    return /* @__PURE__ */ jsxs(
      "li",
      {
        className: clsx(
          breadcrumbsListItemCSS({ isLastVisible }),
          consumerClassName
        ),
        style: consumerStyle,
        "data-id": dataId,
        "data-testid": dataTestId,
        children: [
          /* @__PURE__ */ jsx(
            Link,
            {
              className: clsx(
                focusRingStatic({
                  minimal: true,
                  color: remainingProps.menuTrigger ? "neutral" : "primary",
                  disabled
                }),
                breadcrumbsItemCSS({
                  disabled,
                  isLastItem: isLast
                })
              ),
              ref: forwardedRef,
              as,
              href,
              "aria-disabled": disabled,
              "aria-current": isLast && "page",
              ...remainingProps,
              tabIndex: !disabled ? 0 : -1,
              children
            }
          ),
          !isCollapsed && !isLast && /* @__PURE__ */ jsx(ChevronRightSmallIcon, { className: breadcrumbsSeperatorIconCSS })
        ]
      }
    );
  }
);
Item.displayName = "Breadcrumb.Item";
export {
  Item,
  isBreadcrumbItem
};
