// @dynatrace/strato-components/esm/filters/segment-selector/SegmentSelector.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import {
  Children,
  forwardRef,
  useEffect,
  useImperativeHandle,
  useMemo,
  useRef
} from "react";
import { useIntl } from "react-intl";
import { isElement } from "react-is";
import { FocusScope } from "../../core/components/focus-scope/FocusScope.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useId } from "../../core/hooks/useId.js";
import { useSyncedFormControlProps } from "../../forms/forms-core/hooks/useSyncedFormControlProps.js";
import { FormFieldMessagesAction } from "../../forms/forms-core/state/formfield-messages-state-reducer.js";
import { segmentsClient } from "./contexts/SegmentsCache/SegmentsCache.js";
import { SegmentsTriggerContext } from "./contexts/SegmentsUI/SegmentsTriggerContext.js";
import { UISegmentsProvider } from "./contexts/SegmentsUI/SegmentsUIContext.js";
import { validateSegments } from "./contexts/SegmentsUI/segmentValidationHelper.js";
import { CustomTrigger } from "./CustomTrigger.js";
import { useSegments } from "./hooks/useSegments.js";
import { useTooltipHandler } from "./hooks/useTooltipHandler.js";
import { useTriggerProps } from "./hooks/useTriggerProps.js";
import { SegmentsOverlay } from "./parts/Overlay.js";
import { SegmentSelectorTrigger } from "./parts/selectionBar/SegmentSelectorTrigger.js";
import { TooltipMessages } from "./parts/TooltipMessages.js";
import { SEGMENTSELECTOR_MESSAGES } from "./segmentMessages.js";
import * as styles from "./SegmentSelector.sty.js";
import { OverlayContainer } from "../../core/components/OverlayContainer.js";
import { VisuallyHidden } from "../../core/components/visually-hidden/VisuallyHidden.js";
import { useMergeRefs } from "../../core/hooks/useMergeRefs.js";
import { createStratoLogger } from "../../core/sdk/sdk-logger.js";
import { StratoLoggerContext } from "../../core/sdk/StratoLoggerContext.js";
import { isDev } from "../../core/utils/is-node-environment.js";
import { mergeProps } from "../../core/utils/merge-props.js";
import { withFormFieldMessageContextProvider } from "../../forms/forms-core/components/form-field/FormFieldMessageContextProvider.js";
import { useFormFieldMessagesContext } from "../../forms/forms-core/hooks/useFormFieldMessagesContext.js";
import { useFormFieldValidation } from "../../forms/forms-core/hooks/useFormFieldValidation.js";
import { useOverlayWithTrigger } from "../../overlays/hooks/useOverlayWithTrigger.js";
import { overlayCSS } from "../../overlays/overlay/Overlay.sty.js";
import { Tooltip } from "../../overlays/tooltip/Tooltip.js";
const COMPONENT_NAME = "SegmentSelector";
function getCustomTriggerSlot(children) {
  let customTrigger;
  Children.forEach(children, (child) => {
    if (isElement(child)) {
      switch (child.type) {
        case CustomTrigger:
          customTrigger = child;
          break;
        default: {
          if (isDev()) {
            console.warn(
              "Only CustomTrigger is a valid child of the SegmentSelector."
            );
          }
          break;
        }
      }
    } else if (child && isDev()) {
      console.warn("The used child is not supported by the SegmentSelector.");
    }
  });
  return customTrigger;
}
const SegmentSelectorFwd = forwardRef((props, forwardedRef) => {
  const {
    className: consumerClassName,
    style: consumerStyle,
    id: propId,
    "data-testid": dataTestId,
    "data-dtrum-allow": dataDtrumAllow,
    "data-dtrum-mask": dataDtrumMask,
    variant = "default",
    onOpenChange,
    readOnly: propReadOnly,
    disabled: propDisabled,
    required: propRequired,
    name,
    ...remainingProps
  } = props;
  const intl = useIntl();
  const inputRef = useRef(null);
  const internalButtonRef = useRef(null);
  const { disabled, required, id, ariaDisabled, readOnly } = useSyncedFormControlProps({
    propDisabled,
    propRequired,
    propId,
    propAriaDisabled: remainingProps["aria-disabled"],
    propReadOnly
  });
  const {
    isLoading,
    text,
    ariaProps: triggerAriaProps
  } = useTriggerProps(variant, remainingProps);
  const [behavioralTrackingProps] = useBehavioralTrackingProps(remainingProps);
  const { onChange, onInvalid, ariaProps } = useFormFieldValidation(
    inputRef,
    { required },
    triggerAriaProps,
    id
  );
  const { state: formFieldMessagesState, dispatch: formFieldMessagesDispatch } = useFormFieldMessagesContext();
  const errorId = useId();
  const { segments } = useSegments();
  const stringifiedValue = JSON.stringify(segments);
  const segmentSelectorProps = {
    "data-dtrum-allow": dataDtrumAllow,
    "data-dtrum-mask": dataDtrumMask,
    "data-testid": dataTestId,
    style: consumerStyle,
    className: consumerClassName,
    ...mergeProps(behavioralTrackingProps, ariaProps)
  };
  const customTrigger = useMemo(
    () => getCustomTriggerSlot(remainingProps.children),
    [remainingProps.children]
  );
  const {
    targetRef,
    triggerRef,
    overlayRef,
    overlayState,
    overlayProps: _overlayProps,
    overlayTriggerProps,
    overlayContainerProps
  } = useOverlayWithTrigger({
    placement: "bottom-start",
    placementFallbacks: ["bottom-end", "top-start", "top-end"],
    offset: 4,
    shouldCloseOnBlur: false,
    onOpenChange: (isOpen2) => {
      if (!isOpen2 && !formFieldMessagesState.isTouched && segmentsClient.hasAccessRights !== false) {
        formFieldMessagesDispatch?.({
          type: FormFieldMessagesAction.SETISTOUCHED,
          payload: { isTouched: true }
        });
      }
      if (!isOpen2) {
        validateSegments(
          segments,
          onChange,
          formFieldMessagesDispatch,
          formFieldMessagesState.formControlMessages,
          errorId,
          true
        ).catch(() => void 0);
      }
      onOpenChange?.(isOpen2);
    },
    additionalElevation: "overlay",
    disabled
  });
  const { trapFocus, ...overlayProps } = _overlayProps;
  const { onOpenTooltip, isOpen, textEllipsisRef } = useTooltipHandler(
    !customTrigger && variant === "default"
  );
  const overlayAriaLabelledById = useId();
  useImperativeHandle(
    forwardedRef,
    () => ({
      element: internalButtonRef.current,
      triggerRef: internalButtonRef.current,
      inputRef: inputRef.current,
      focus: () => internalButtonRef.current?.focus(),
      open: overlayState.open,
      close: overlayState.close,
      validate: async () => {
        formFieldMessagesDispatch?.({
          type: FormFieldMessagesAction.SETISTOUCHED,
          payload: { isTouched: true }
        });
        if (segmentsClient.hasAccessRights === false) {
          formFieldMessagesDispatch?.({
            type: FormFieldMessagesAction.ADDMESSAGE,
            payload: {
              id: errorId,
              variant: "error",
              validityStateKey: "customError",
              messageType: "formcontrol",
              showError: true,
              message: intl.formatMessage(
                SEGMENTSELECTOR_MESSAGES.noReadAndWriteAccess
              )
            }
          });
        }
        validateSegments(
          segments,
          onChange,
          formFieldMessagesDispatch,
          formFieldMessagesState.formControlMessages,
          errorId,
          true
        ).catch(() => void 0);
      }
    }),
    [
      errorId,
      formFieldMessagesDispatch,
      formFieldMessagesState.formControlMessages,
      intl,
      onChange,
      overlayState.close,
      overlayState.open,
      segments
    ]
  );
  const buttonRef = useMergeRefs([triggerRef, targetRef, internalButtonRef]);
  useEffect(() => {
    validateSegments(
      segments,
      onChange,
      formFieldMessagesDispatch,
      formFieldMessagesState.formControlMessages,
      errorId
    ).catch(() => void 0);
  }, [
    errorId,
    formFieldMessagesDispatch,
    formFieldMessagesState.formControlMessages,
    onChange,
    segments,
    stringifiedValue
  ]);
  const providerValue = useMemo(
    () => ({
      isLoading,
      displayValue: text,
      triggerProps: {
        ...overlayTriggerProps,
        "data-dt-component": COMPONENT_NAME
      },
      buttonRef,
      showError: formFieldMessagesState.showError,
      ...ariaProps
    }),
    [
      isLoading,
      text,
      overlayTriggerProps,
      buttonRef,
      formFieldMessagesState.showError,
      ariaProps
    ]
  );
  const stratoLogger = useMemo(() => {
    return createStratoLogger.bind({
      name,
      id: propId,
      ariaLabel: triggerAriaProps["aria-label"]
    })(COMPONENT_NAME);
  }, [name, propId, triggerAriaProps]);
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsxs(SegmentsTriggerContext.Provider, { value: providerValue, children: [
      /* @__PURE__ */ jsx(
        Tooltip,
        {
          text: /* @__PURE__ */ jsx(TooltipMessages, { ariaLabel: triggerAriaProps["aria-label"] }),
          className: styles.overwriteTooltipElementPadding,
          onOpenChange: onOpenTooltip,
          open: isOpen,
          children: customTrigger ?? /* @__PURE__ */ jsx(
            SegmentSelectorTrigger,
            {
              textEllipsisRef,
              ...segmentSelectorProps
            }
          )
        }
      ),
      /* @__PURE__ */ jsx(VisuallyHidden, { children: /* @__PURE__ */ jsx(
        "input",
        {
          ref: inputRef,
          id,
          value: segments.length > 0 ? stringifiedValue : "",
          onChange: () => void 0,
          "aria-hidden": true,
          readOnly,
          disabled,
          name,
          required,
          tabIndex: -1,
          "aria-disabled": ariaDisabled,
          onInvalid
        }
      ) })
    ] }),
    overlayState.isOpen && /* @__PURE__ */ jsx(StratoLoggerContext.Provider, { value: stratoLogger, children: /* @__PURE__ */ jsx(
      OverlayContainer,
      {
        "data-testid": dataTestId,
        ...overlayContainerProps,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        children: /* @__PURE__ */ jsx(FocusScope, { autoFocus: true, contain: trapFocus, children: /* @__PURE__ */ jsx(
          "div",
          {
            ref: overlayRef,
            className: overlayCSS,
            ...overlayProps,
            "aria-labelledby": overlayAriaLabelledById,
            children: /* @__PURE__ */ jsx(
              UISegmentsProvider,
              {
                readOnly,
                overlayState,
                children: /* @__PURE__ */ jsx(SegmentsOverlay, { labelId: overlayAriaLabelledById })
              }
            )
          }
        ) })
      }
    ) })
  ] });
});
const SegmentSelectorRoot = withFormFieldMessageContextProvider(SegmentSelectorFwd);
const SegmentSelector = Object.assign(SegmentSelectorRoot, {
  CustomTrigger
});
SegmentSelector.displayName = COMPONENT_NAME;
export {
  SegmentSelector
};
