// @dynatrace/strato-components/layouts/page-layout/sidebar-layout/CollapsedSidebarRail.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CollapsedSidebarRail_exports = {};
__export(CollapsedSidebarRail_exports, {
  SidebarHint: () => SidebarHint
});
module.exports = __toCommonJS(CollapsedSidebarRail_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_strato_icons = require("@dynatrace/strato-icons");
var styles = __toESM(require("./SidebarLayout.sty.js"));
var import_Button = require("../../../buttons/button/Button.js");
var import_Tooltip = require("../../../overlays/tooltip/Tooltip.js");
function SidebarHint({
  panelId,
  expandSidebarLabel,
  expandButtonRef,
  onCollapsedChange,
  isCollapsed = true
}) {
  const expand = () => {
    onCollapsedChange(false);
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { className: styles.collapsedSidebar, id: panelId, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "button",
      {
        className: styles.hint,
        type: "button",
        onClick: expand,
        tabIndex: -1,
        "aria-hidden": "true",
        "aria-label": expandSidebarLabel
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: styles.expandButtonContainer, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Tooltip.Tooltip, { text: expandSidebarLabel, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Button.Button,
      {
        className: styles.expandButton,
        "aria-expanded": !isCollapsed,
        ref: expandButtonRef,
        "aria-controls": panelId,
        "aria-label": expandSidebarLabel,
        size: "condensed",
        "data-dt-name": "ExpandCollapse",
        "data-dt-properties": "state:expand",
        onClick: expand,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ChevronSmallRightDoubleIcon, {}) })
      }
    ) }) })
  ] });
}
