// @dynatrace/strato-components/layouts/page-layout/sidebar-layout/SidebarLayout.sty.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SidebarLayout_css_exports = {};
__export(SidebarLayout_css_exports, {
  collapseButton: () => collapseButton,
  collapseButtonContainer: () => collapseButtonContainer,
  collapsedSidebar: () => collapsedSidebar,
  collapsedSidebarWrapper: () => collapsedSidebarWrapper,
  content: () => content,
  drawer: () => drawer,
  drawerContent: () => drawerContent,
  drawerContentContainer: () => drawerContentContainer,
  drawerHeader: () => drawerHeader,
  expandButton: () => expandButton,
  expandButtonContainer: () => expandButtonContainer,
  hint: () => hint,
  layoutRoot: () => layoutRoot,
  panel: () => panel,
  resizeHandle: () => resizeHandle
});
module.exports = __toCommonJS(SidebarLayout_css_exports);
var import_AppRoot_css_ts_vanilla = require("../../../core/components/app-root/AppRoot.css");
var import_focusRingStatic_css_ts_vanilla = require("../../../core/styles/focusRingStatic.css");
var import_SidebarLayout_css_ts_vanilla = require("./SidebarLayout.css");
var collapseButton = "_1mr4xg7f-3-6-0 _1mr4xg7e-3-6-0";
var collapseButtonContainer = "_1mr4xg7c-3-6-0";
var collapsedSidebar = "_1mr4xg75-3-6-0";
var collapsedSidebarWrapper = "_1mr4xg71-3-6-0";
var content = "_1mr4xg73-3-6-0";
var drawer = "_1mr4xg77-3-6-0";
var drawerContent = "_1mr4xg79-3-6-0";
var drawerContentContainer = "_1mr4xg78-3-6-0";
var drawerHeader = "_1mr4xg7a-3-6-0";
var expandButton = "_1mr4xg7g-3-6-0 _1mr4xg7e-3-6-0";
var expandButtonContainer = "_1mr4xg7d-3-6-0";
var hint = "_1mr4xg76-3-6-0 _z0fq9b4-3-6-0 _z0fq9bd-3-6-0 _z0fq9b6-3-6-0 _z0fq9bc-3-6-0";
var layoutRoot = "_1mr4xg70-3-6-0";
var panel = "_1mr4xg72-3-6-0";
var resizeHandle = "_1mr4xg74-3-6-0";
