// @dynatrace/strato-components/layouts/page-layout/sidebar-layout/SidebarLayout.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SidebarLayout_exports = {};
__export(SidebarLayout_exports, {
  SidebarLayout: () => SidebarLayout
});
module.exports = __toCommonJS(SidebarLayout_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = require("clsx");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_CollapsedSidebarRail = require("./CollapsedSidebarRail.js");
var styles = __toESM(require("./SidebarLayout.sty.js"));
var import_useFocusHandover = require("./useFocusHandover.js");
var import_useLayoutState = require("./useLayoutState.js");
var import_Button = require("../../../buttons/button/Button.js");
var import_useId = require("../../../core/hooks/useId.js");
var import_useMergeRefs = require("../../../core/hooks/useMergeRefs.js");
var import_Drawer = require("../../../overlays/drawer/Drawer.js");
var import_Tooltip = require("../../../overlays/tooltip/Tooltip.js");
var import_SplitLayout = require("../../split-layout/SplitLayout.js");
var import_useTeleport = require("../shared/useTeleport.js");
function SidebarLayoutInner({
  sidebar,
  content,
  containerRef,
  panelId,
  persistedPanelWidthRef,
  contentTeleport,
  panelTeleport
}) {
  const layout = (0, import_useLayoutState.useLayoutState)({ sidebar, containerRef });
  const intl = (0, import_react_intl.useIntl)();
  const focusHandover = (0, import_useFocusHandover.useButtonFocusHandover)();
  const expandSidebarLabel = intl.formatMessage({
    defaultMessage: "Expand sidebar",
    id: "NF0tLfJTokD/xw5B",
    description: "Label for the button that expands the sidebar."
  });
  const collapseSidebarLabel = intl.formatMessage({
    defaultMessage: "Collapse sidebar",
    id: "LJ0fyD2++VZgZ5g0",
    description: "Label for the button that collapses the sidebar."
  });
  const sidebarLabel = intl.formatMessage({
    defaultMessage: "Sidebar",
    id: "BYUJpW8TBonAxnYd",
    description: "Accessible label for the sidebar panel."
  });
  const setCollapsedWithFocusHandover = (value) => {
    layout.setCollapsed(value);
    focusHandover.setTarget(value ? "expand" : "collapse");
  };
  switch (layout.mode) {
    case "collapsed": {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { className: styles.collapsedSidebarWrapper, children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_CollapsedSidebarRail.SidebarHint,
          {
            panelId,
            expandSidebarLabel,
            expandButtonRef: focusHandover.expandButtonRef,
            onCollapsedChange: setCollapsedWithFocusHandover
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          "div",
          {
            ref: content.ref,
            ...content.domProps,
            className: (0, import_clsx.clsx)(styles.content, content.domProps.className),
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(contentTeleport.Target, {})
          }
        )
      ] });
    }
    case "split": {
      const splitLayoutKey = persistedPanelWidthRef.current === void 0 ? `split-${String(sidebar.config.defaultWidth)}` : "split-resized";
      return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_SplitLayout.SplitLayout, { children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_SplitLayout.SplitLayout.Panel,
          {
            ...sidebar.domProps,
            ref: sidebar.ref,
            minWidth: sidebar.config.minWidth,
            maxWidth: sidebar.config.maxWidth,
            defaultWidth: persistedPanelWidthRef.current ?? sidebar.config.defaultWidth,
            resizable: sidebar.config.resizable,
            position: "start",
            onResize: (value) => {
              persistedPanelWidthRef.current = value;
              sidebar.config.onResize?.(value);
            },
            className: (0, import_clsx.clsx)(styles.panel, sidebar.domProps.className),
            "aria-label": sidebarLabel,
            id: panelId,
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(panelTeleport.Target, {})
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_SplitLayout.SplitLayout.ResizeHandle, { className: styles.resizeHandle, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: styles.collapseButtonContainer, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Tooltip.Tooltip, { text: collapseSidebarLabel, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_Button.Button,
          {
            className: styles.collapseButton,
            "aria-expanded": "true",
            ref: focusHandover.collapseButtonRef,
            "aria-controls": panelId,
            "aria-label": collapseSidebarLabel,
            size: "condensed",
            "data-dt-name": "ExpandCollapse",
            "data-dt-properties": "state:collapse",
            onClick: () => {
              setCollapsedWithFocusHandover(true);
            },
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ChevronSmallLeftDoubleIcon, {}) })
          }
        ) }) }) }),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_SplitLayout.SplitLayout.Content,
          {
            ...content.domProps,
            ref: content.ref,
            minWidth: content.config.minWidth,
            className: (0, import_clsx.clsx)(styles.content, content.domProps.className),
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(contentTeleport.Target, {})
          }
        )
      ] }, splitLayoutKey);
    }
    case "drawer": {
      const drawerContentArea = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        "div",
        {
          ref: content.ref,
          ...content.domProps,
          className: (0, import_clsx.clsx)(styles.content, content.domProps.className),
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(contentTeleport.Target, {})
        }
      );
      return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { className: styles.collapsedSidebarWrapper, children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_CollapsedSidebarRail.SidebarHint,
            {
              panelId,
              expandSidebarLabel,
              expandButtonRef: focusHandover.expandButtonRef,
              isCollapsed: layout.collapsed,
              onCollapsedChange: setCollapsedWithFocusHandover
            }
          ),
          drawerContentArea
        ] }),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_Drawer.Drawer,
          {
            ...sidebar.domProps,
            ref: sidebar.ref,
            className: (0, import_clsx.clsx)(styles.drawer, sidebar.domProps.className),
            "aria-label": sidebarLabel,
            isDismissed: layout.collapsed,
            onDismiss: (state, _reason) => {
              setCollapsedWithFocusHandover(state);
            },
            placement: "left",
            width: "wide",
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { className: styles.drawerContentContainer, children: [
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: styles.drawerHeader, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_Button.Button,
                {
                  style: { marginLeft: "auto" },
                  ref: focusHandover.collapseButtonRef,
                  "aria-label": collapseSidebarLabel,
                  "data-dt-name": "CloseX",
                  onClick: () => {
                    setCollapsedWithFocusHandover(true);
                  },
                  children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.XmarkIcon, {}) })
                }
              ) }),
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: styles.drawerContent, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(panelTeleport.Target, {}) })
            ] })
          }
        )
      ] });
    }
  }
}
const SidebarLayout = (0, import_react.memo)(
  (0, import_react.forwardRef)((props, ref) => {
    const { className: rootClassName, sidebar, content, ...rootProps } = props;
    const panelId = (0, import_useId.useId)();
    const containerRef = (0, import_react.useRef)(null);
    const rootRef = (0, import_useMergeRefs.useMergeRefs)([ref, containerRef]);
    const persistedPanelWidth = (0, import_react.useRef)();
    const contentTeleport = (0, import_useTeleport.useTeleport)();
    const panelTeleport = (0, import_useTeleport.useTeleport)();
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(contentTeleport.Source, { children: content.children }),
      sidebar && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(panelTeleport.Source, { children: sidebar.children }),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        "div",
        {
          ...rootProps,
          ref: rootRef,
          className: (0, import_clsx.clsx)(styles.layoutRoot, rootClassName),
          children: sidebar ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            SidebarLayoutInner,
            {
              sidebar,
              content,
              containerRef,
              panelId,
              persistedPanelWidthRef: persistedPanelWidth,
              contentTeleport,
              panelTeleport
            }
          ) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "div",
            {
              ref: content.ref,
              ...content.domProps,
              className: (0, import_clsx.clsx)(styles.content, content.domProps.className),
              children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(contentTeleport.Target, {})
            }
          )
        }
      )
    ] });
  })
);
SidebarLayout.displayName = "SidebarLayout (forwardRef)";
