// @dynatrace/strato-components/forms/date-time-range-picker/DateTimeInput.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DateTimeInput_exports = {};
__export(DateTimeInput_exports, {
  DateTimeInput: () => DateTimeInput
});
module.exports = __toCommonJS(DateTimeInput_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_DateTimeInput_css = require("./DateTimeInput.sty.js");
var styles = __toESM(require("./DateTimeInput.sty.js"));
var import_useDateTimeInput = require("./hooks/useDateTimeInput.js");
var visuallyHiddenStyles = __toESM(require("../../core/components/visually-hidden/VisuallyHidden.sty.js"));
var import_VisuallyHidden = require("../../core/components/visually-hidden/VisuallyHidden.js");
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useFormControlState = require("../../core/hooks/useFormControlState.js");
var import_useTimeframeStore = require("../../core/time/timeframeStore/useTimeframeStore.js");
var import_validation = require("../../core/time/validation.js");
var import_merge_props = require("../../core/utils/merge-props.js");
var import_Flex = require("../../layouts/flex/Flex.js");
var dataTimePickerStyles = __toESM(require("../date-time-picker/DateTimePicker.sty.js"));
var import_useFormFieldMessagesContext = require("../forms-core/hooks/useFormFieldMessagesContext.js");
var import_useFormFieldValidation = require("../forms-core/hooks/useFormFieldValidation.js");
var import_formfield_messages_state_reducer = require("../forms-core/state/formfield-messages-state-reducer.js");
var import_useInputHint = require("../hooks/useInputHint.js");
var import_SpinButtonGroup = require("../spin-buttons/spin-button-group/SpinButtonGroup.js");
function preventDefault(e) {
  e.preventDefault();
}
const DateTimeInput = (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const rootRef = (0, import_react.useRef)(null);
    const inputRef = (0, import_react.useRef)(null);
    const expressionInputRef = (0, import_react.useRef)(null);
    const min = (0, import_useTimeframeStore.useDateTimePickerStore)((state) => state.min);
    const max = (0, import_useTimeframeStore.useDateTimePickerStore)((state) => state.max);
    const precision = (0, import_useTimeframeStore.useDateTimePickerStore)((state) => state.precision);
    const formatString = (0, import_useTimeframeStore.useDateTimePickerStore)((state) => state.formatString);
    const intl = (0, import_react_intl.useIntl)();
    const {
      ariaProps: ariaFormProps,
      onChange: validationOnChange,
      onBlur,
      onInvalid
    } = (0, import_useFormFieldValidation.useFormFieldValidation)(inputRef);
    const {
      disabled,
      readOnly,
      placeholder: propPlaceholder,
      className: consumerClassName,
      style: consumerStyle,
      name,
      "data-testid": dataTestId,
      "aria-disabled": ariaDisabledProp,
      controlId,
      required,
      type = "datetime",
      controlState,
      validationValueMode,
      onLeadingEdge,
      onTrailingEdge,
      onTrailingOverflow
    } = props;
    const dateTimePlaceholder = intl.formatMessage({
      defaultMessage: "Add date and time",
      id: "X9SZ3eLw4rz/ODZQ",
      description: "Default placeholder for the date time input."
    });
    const datePlaceholder = intl.formatMessage({
      defaultMessage: "Add date",
      id: "XC7nO0JIPg0R0kBw",
      description: "Default placeholder for the date input."
    });
    const timePlaceholder = intl.formatMessage({
      defaultMessage: "Add time",
      id: "MlwM+/jonCtjJ+L7",
      description: "Default placeholder for the time input."
    });
    let placeholder = propPlaceholder;
    if (!propPlaceholder) {
      switch (type) {
        case "datetime":
          placeholder = dateTimePlaceholder;
          break;
        case "date":
          placeholder = datePlaceholder;
          break;
        case "time":
          placeholder = timePlaceholder;
          break;
      }
    }
    const [hasFormControlError, hint] = (0, import_useFormControlState.useFormControlState)(controlState);
    const { inputProps } = (0, import_useInputHint.useInputHint)({
      hint,
      hasError: hasFormControlError
    });
    const ariaLabelingProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(props);
    const {
      expressionInputHandler,
      expressionOnFocusHandler,
      expressionOnBlurHandler,
      expressionOnKeyDownHandler,
      onPaste,
      pasteText,
      onPasteUndoKeyDown,
      clearPasteUndo,
      value,
      valueType,
      spinButtons,
      spinButtonHandlers,
      inputValue,
      focus,
      switchToExpressionMode,
      setInputValue
    } = (0, import_useDateTimeInput.useDateTimeInput)(props, expressionInputRef);
    const isExpressionInput = valueType === "expression";
    const { dispatch: formMessageDispatch } = (0, import_useFormFieldMessagesContext.useFormFieldMessagesContext)();
    (0, import_react.useImperativeHandle)(
      forwardedRef,
      () => ({
        element: rootRef.current,
        inputRef: inputRef.current,
        expressionInputRef: expressionInputRef.current,
        spinButtonHandlers,
        focus,
        setExpressionValue: switchToExpressionMode,
        clearPasteUndo,
        pasteText,
        focusLast() {
          if (valueType === "expression") {
            expressionInputRef.current?.focus();
          } else {
            const handlers = Array.from(spinButtonHandlers.values()).filter(
              (h) => h !== null
            );
            const lastHandler = handlers[handlers.length - 1];
            lastHandler.element.current?.focus();
            lastHandler.selectAll();
          }
        },
        validate() {
          const textContent = expressionInputRef.current?.textContent;
          if (valueType === "expression" && !textContent) {
            setInputValue("");
          }
          (0, import_validation.validate)(value ?? null, min, max, {
            formMessageDispatch,
            precision,
            showError: true,
            checkOrder: true,
            ref: rootRef,
            expectedFormat: formatString,
            valueMode: validationValueMode,
            // In single mode, required validation is handled by core time validation
            // (singleValueMissing). Keeping native `required` here would add a second
            // browser `valueMissing` message with identical text.
            required: validationValueMode === "single" ? false : required
          });
          formMessageDispatch?.({
            type: import_formfield_messages_state_reducer.FormFieldMessagesAction.SETISTOUCHED,
            payload: { isTouched: true }
          });
        }
      }),
      [
        spinButtonHandlers,
        focus,
        switchToExpressionMode,
        setInputValue,
        value,
        valueType,
        min,
        max,
        formMessageDispatch,
        precision,
        formatString,
        validationValueMode,
        required
      ]
    );
    const expressionLabel = intl.formatMessage(
      {
        defaultMessage: "{ariaLabelPrefix} expression",
        id: "M6eNgatb2wf4fQ3I",
        description: "Textbox to enter a dql timeframe expression."
      },
      {
        ariaLabelPrefix: ariaLabelingProps["aria-label"]
      }
    );
    (0, import_react.useEffect)(() => {
      validationOnChange();
    }, [inputValue]);
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      import_Flex.Flex,
      {
        ref: rootRef,
        as: "span",
        className: (0, import_clsx.default)(
          import_DateTimeInput_css.dateTimeInputRoot,
          styles.dateTimeInput,
          consumerClassName
        ),
        style: consumerStyle,
        "data-testid": dataTestId,
        onPaste,
        onKeyDown: onPasteUndoKeyDown,
        gap: 0,
        "data-valuetype": valueType,
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_VisuallyHidden.VisuallyHidden, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "input",
            {
              id: controlId,
              ref: inputRef,
              name,
              value: inputValue ?? "",
              onChange: () => {
              },
              onFocus: focus,
              onBlur,
              onInvalid,
              disabled,
              tabIndex: -1,
              required,
              readOnly,
              "aria-disabled": ariaDisabledProp,
              ...(0, import_merge_props.mergeProps)(
                // Backwards compatibility for controlState.
                controlState ? inputProps : ariaFormProps,
                ariaLabelingProps
              )
            }
          ) }),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "span",
            {
              className: (0, import_clsx.default)(import_DateTimeInput_css.expression, {
                [visuallyHiddenStyles.visuallyHidden]: !isExpressionInput
              }),
              role: "textbox",
              spellCheck: false,
              tabIndex: disabled || !isExpressionInput ? -1 : 0,
              style: {
                "--data-placeholder": `"${placeholder}"`
              },
              "data-valuetype": "expression",
              ref: expressionInputRef,
              contentEditable: !disabled && !readOnly ? true : void 0,
              onInput: expressionInputHandler,
              onFocus: (e) => {
                if (e.target.hasAttribute("contenteditable") === false) {
                  return;
                }
                expressionOnFocusHandler(e);
              },
              onBlur: (e) => {
                if (e.target.hasAttribute("contenteditable") === false) {
                  return;
                }
                expressionOnBlurHandler(e);
              },
              onKeyDown: expressionOnKeyDownHandler,
              onPaste: preventDefault,
              onDragOver: preventDefault,
              onDrop: preventDefault,
              "aria-label": expressionLabel
            }
          ),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_SpinButtonGroup.SpinButtonGroup,
            {
              className: (0, import_clsx.default)(dataTimePickerStyles.spinButtonGroup, {
                [visuallyHiddenStyles.visuallyHidden]: isExpressionInput
              }),
              valueOverflow: true,
              onLeadingEdge,
              onTrailingEdge,
              onTrailingOverflow,
              children: spinButtons
            }
          )
        ]
      }
    );
  }
);
