// @dynatrace/strato-components/forms/date-time-range-picker/DateTimeRangePicker.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DateTimeRangePicker_exports = {};
__export(DateTimeRangePicker_exports, {
  DateTimeRangePicker: () => DateTimeRangePicker
});
module.exports = __toCommonJS(DateTimeRangePicker_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_DateTimeRangePickerRoot = require("./DateTimeRangePickerRoot.js");
var import_useCallbackRef = require("../../core/hooks/useCallbackRef.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_dateTimeRangePicker = require("../../core/time/timeframeStore/actions/dateTimeRangePicker.js");
var import_createTimeframeStore = require("../../core/time/timeframeStore/createTimeframeStore.js");
var import_timeframeStoreProvider = require("../../core/time/timeframeStore/timeframeStoreProvider.js");
var import_useDateTimeMessages = require("../../core/time/useDateTimeMessages.js");
var import_useMinMax = require("../../core/time/useMinMax.js");
var import_validation = require("../../core/time/validation.js");
var import_helpers = require("../calendar/utils/helpers.js");
var import_useFormFieldMessagesContext = require("../forms-core/hooks/useFormFieldMessagesContext.js");
var import_useSyncedFormControlProps = require("../forms-core/hooks/useSyncedFormControlProps.js");
var import_formfield_messages_state_reducer = require("../forms-core/state/formfield-messages-state-reducer.js");
var import_date_time = require("../forms-core/time/date-time.js");
var import_timeParser = require("../forms-core/time/timeParser.js");
function useSyncPropsAndState(store, memoedValue, isControlled, precision, formatString, required, formFieldMessagesDispatch, dateTimeRangePickerRef) {
  (0, import_react.useEffect)(() => {
    isControlled && store.setState((state) => {
      const error = (0, import_validation.validate)(memoedValue ?? null, state.min, state.max, {
        formMessageDispatch: formFieldMessagesDispatch,
        precision: state.precision,
        showError: false,
        checkOrder: true,
        formerError: state.calendarError,
        ref: dateTimeRangePickerRef,
        expectedFormat: state.formatString,
        valueMode: "dateTimePicker",
        required: state.required
      });
      const value = (0, import_timeParser.removePrecisionFromTimeFrame)(
        memoedValue,
        state.precision
      );
      const timeOffset = (0, import_helpers.getTimeOffsetFromInitialValue)(
        value,
        "range",
        state.min,
        state.max,
        precision
      );
      return {
        ...state,
        dateTimeRangePickerTimeframe: value,
        calendarValue: value,
        calendarError: error,
        dateTimeRangePickerError: error,
        timeOffset
      };
    });
  }, [
    formFieldMessagesDispatch,
    isControlled,
    memoedValue,
    store,
    precision,
    dateTimeRangePickerRef
  ]);
  (0, import_react.useEffect)(() => {
    store.setState((state) => {
      const value = (0, import_timeParser.removePrecisionFromTimeFrame)(
        state.dateTimeRangePickerTimeframe,
        precision
      );
      return {
        ...state,
        precision,
        formatString,
        dateTimeRangePickerTimeframe: value,
        calendarValue: value,
        required
      };
    });
  }, [precision, formatString, store, required]);
}
const DateTimeRangePicker = (0, import_react.forwardRef)(
  ({
    value: propValue,
    defaultValue: propDefaultValue,
    onChange,
    min: propMin = import_date_time.GRAIL_MIN_DATE,
    max: propMax = import_date_time.GRAIL_MAX_DATE,
    precision = "minutes",
    disabled: propDisabled,
    required: propRequired,
    readOnly: propReadOnly,
    id: propId,
    "aria-disabled": propAriaDisabled,
    onBlur,
    ...remainingProps
  }, forwardedRef) => {
    const [isControlled] = (0, import_react.useState)(propValue !== void 0);
    const onChangeProp = (0, import_useCallbackRef.useCallbackRef)(onChange);
    const memoedValue = (0, import_react.useMemo)(
      () => (0, import_timeParser.mapToPartialTimeframe)(propValue, precision),
      [precision, propValue]
    );
    const memoedDefaultValue = (0, import_react.useMemo)(
      () => (0, import_timeParser.mapToPartialTimeframe)(propDefaultValue, precision),
      /** We're ignoring updates to defaultValue as it only considers the initial state. */
      // eslint-disable-next-line react-hooks/exhaustive-deps
      []
    );
    const { min, max } = (0, import_useMinMax.useMinMax)(propMin, propMax);
    const calendarRef = (0, import_react.useRef)(null);
    const dateTimeFromRef = (0, import_react.useRef)(null);
    const dateTimeRangePickerRef = (0, import_react.useRef)(null);
    const { dispatch: formFieldMessagesDispatch } = (0, import_useFormFieldMessagesContext.useFormFieldMessagesContext)();
    const { formatString } = (0, import_useDateTimeMessages.useDateTimeMessages)({ precision });
    const { disabled, required, id, ariaDisabled, readOnly } = (0, import_useSyncedFormControlProps.useSyncedFormControlProps)({
      propDisabled,
      propRequired,
      propId,
      propAriaDisabled,
      propReadOnly
    });
    const [initialState] = (0, import_react.useState)(() => {
      const value = isControlled ? memoedValue : memoedDefaultValue;
      const timeOffset = (0, import_helpers.getTimeOffsetFromInitialValue)(
        value,
        "range",
        min,
        max,
        precision
      );
      const error = (0, import_validation.validate)(value, min, max, {
        formMessageDispatch: formFieldMessagesDispatch,
        precision,
        showError: false,
        checkOrder: true,
        ref: dateTimeRangePickerRef,
        expectedFormat: formatString,
        valueMode: "dateTimePicker",
        required
      });
      return {
        calendarValue: value,
        dateTimeRangePickerTimeframe: value,
        min,
        max,
        precision,
        calendarRef,
        dateTimeFromRef,
        dateTimeRangePickerRef,
        calendarError: error,
        dateTimeRangePickerError: error,
        timeOffset,
        formatString,
        required
      };
    });
    const [store] = (0, import_react.useState)(
      () => (0, import_createTimeframeStore.createTimeframeStore)(
        initialState,
        (0, import_dateTimeRangePicker.getDateTimeRangePickerActions)(formFieldMessagesDispatch, onChangeProp)
      )
    );
    useSyncPropsAndState(
      store,
      memoedValue,
      isControlled,
      precision,
      formatString,
      required,
      formFieldMessagesDispatch,
      dateTimeRangePickerRef
    );
    (0, import_useMinMax.useMinMaxStoreSync)(
      store,
      min,
      max,
      (0, import_react.useCallback)(
        (state, min2, max2) => {
          const error = (0, import_validation.validate)(state.dateTimeRangePickerTimeframe, min2, max2, {
            formMessageDispatch: formFieldMessagesDispatch,
            formerError: state.calendarError,
            precision: state.precision,
            showError: false,
            checkOrder: true,
            ref: dateTimeRangePickerRef,
            expectedFormat: state.formatString,
            valueMode: "dateTimePicker",
            required: state.required
          });
          return { calendarError: error, dateTimeRangePickerError: error };
        },
        [formFieldMessagesDispatch]
      )
    );
    const mergedRef = (0, import_useMergeRefs.useMergeRefs)([
      dateTimeRangePickerRef,
      forwardedRef
    ]);
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_timeframeStoreProvider.DateTimeRangePickerStoreProvider, { store, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_DateTimeRangePickerRoot.DateTimeRangePickerRoot,
      {
        ...remainingProps,
        id,
        ref: mergedRef,
        "aria-disabled": ariaDisabled,
        disabled,
        required,
        readOnly,
        onBlur: (event) => {
          onBlur?.(event);
          if (!event.target.contains(document.activeElement)) {
            formFieldMessagesDispatch?.({
              type: import_formfield_messages_state_reducer.FormFieldMessagesAction.SETISTOUCHED,
              payload: { isTouched: true }
            });
          }
        }
      }
    ) });
  }
);
