// @dynatrace/strato-components/forms/date-time-range-picker/DateRangeSelector.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DateRangeSelector_exports = {};
__export(DateRangeSelector_exports, {
  DateRangeSelector: () => DateRangeSelector
});
module.exports = __toCommonJS(DateRangeSelector_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var styles = __toESM(require("./DateRangeSelector.sty.js"));
var import_Button = require("../../buttons/button/Button.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_useFocusRing = require("../../core/styles/useFocusRing.js");
var import_useTimeframeStore = require("../../core/time/timeframeStore/useTimeframeStore.js");
var import_is_visual_refresh_enabled = require("../../core/utils/is-visual-refresh-enabled.js");
var import_useOverlayWithTrigger = require("../../overlays/hooks/useOverlayWithTrigger.js");
var import_Overlay = require("../../overlays/overlay/Overlay.js");
var import_safe_sprinkles = require("../../styles/safe-sprinkles.js");
var import_CalendarRoot = require("../calendar/CalendarRoot.js");
var import_FormFieldMessageContext = require("../forms-core/contexts/FormFieldMessageContext.js");
const DateRangeSelector = (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    disabled,
    readOnly,
    error,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "aria-disabled": ariaDisabledProp,
    ...remainingProps
  } = props;
  const [open, setOpen] = (0, import_react.useState)(false);
  const intl = (0, import_react_intl.useIntl)();
  const calendarTriggerLabel = intl.formatMessage({
    defaultMessage: "DateRangePicker calendar trigger",
    id: "K2M82PM/gcTo3fjq",
    description: "Text shown for the datetimerangepicker calendar trigger."
  });
  const clearCalendarValue = (0, import_useTimeframeStore.useDateTimeRangePickerStore)(
    (state) => state.clearCalendarValue
  );
  const {
    targetRef,
    triggerRef,
    overlayRef,
    overlayState,
    overlayProps,
    overlayTriggerProps
  } = (0, import_useOverlayWithTrigger.useOverlayWithTrigger)({
    placement: "bottom-end",
    offset: 8,
    shouldCloseOnBlur: false,
    open,
    onOpenChange: (isOpen) => {
      if (!isOpen) {
        clearCalendarValue?.();
      }
      setOpen(isOpen);
    },
    // the overlay should not open when disabled or readOnly
    disabled: disabled || readOnly
  });
  const rootRef = (0, import_useMergeRefs.useMergeRefs)([
    triggerRef,
    targetRef
  ]);
  (0, import_react.useImperativeHandle)(
    forwardedRef,
    () => ({
      open: () => setOpen(true),
      close: () => setOpen(false)
    }),
    []
  );
  const { focusClassName } = (0, import_useFocusRing.useFocusRing)({
    focusWithin: false,
    isMinimal: false,
    disabled,
    color: error ? "critical" : "neutral"
  });
  const formFieldMessageContextValue = (0, import_react.useMemo)(
    () => ({
      dispatch: import_lodash_es.noop,
      state: {
        customMessages: [],
        formControlMessages: [],
        isDisabled: disabled,
        showError: error
      }
    }),
    [error, disabled]
  );
  const isCritical = (0, import_is_visual_refresh_enabled.isVisualRefreshEnabled)() ? error && !disabled && !readOnly : error;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Button.Button,
      {
        ref: rootRef,
        "data-testid": dataTestId,
        className: (0, import_clsx.default)(styles.selector, focusClassName, consumerClassName),
        style: consumerStyle,
        "aria-label": calendarTriggerLabel,
        ...remainingProps,
        ...overlayTriggerProps,
        size: "condensed",
        color: isCritical ? "critical" : "neutral",
        disabled: disabled || readOnly,
        "aria-disabled": ariaDisabledProp,
        variant: "emphasized",
        "data-dt-name": "ShowCalendar",
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.CalendarIcon, {}) })
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Overlay.Overlay,
      {
        ref: overlayRef,
        isOpen: overlayState.isOpen,
        overlayProps,
        className: (0, import_safe_sprinkles.globalSprinkles)({
          paddingLeft: 16,
          paddingTop: 16,
          paddingRight: 16,
          paddingBottom: 16
        }),
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_FormFieldMessageContext.FormFieldMessageContext.Provider, { value: formFieldMessageContextValue, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_CalendarRoot.CalendarRoot, { type: "range", "aria-disabled": ariaDisabledProp }) })
      }
    )
  ] });
});
