// @dynatrace/strato-components/forms/date-time-range-picker/DateTimeRangePickerRoot.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DateTimeRangePickerRoot_exports = {};
__export(DateTimeRangePickerRoot_exports, {
  DateTimeRangePickerRoot: () => DateTimeRangePickerRoot
});
module.exports = __toCommonJS(DateTimeRangePickerRoot_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_DateRangeSelector = require("./DateRangeSelector.js");
var import_DateTimeInput = require("./DateTimeInput.js");
var styles = __toESM(require("./DateTimeRangePicker.sty.js"));
var import_expression_trigger_helpers = require("./expression-trigger-helpers.js");
var import_useDateTimeRangePicker = require("./hooks/useDateTimeRangePicker.js");
var import_selection_copy_helpers = require("./selection-copy-helpers.js");
var import_selection_helpers = require("./selection-helpers.js");
var import_useSelectionInteractionHandlers = require("./useSelectionInteractionHandlers.js");
var import_Button = require("../../buttons/button/Button.js");
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useFocusRing = require("../../core/styles/useFocusRing.js");
var import_useTimeframeStore = require("../../core/time/timeframeStore/useTimeframeStore.js");
var import_validation = require("../../core/time/validation.js");
var import_Flex = require("../../layouts/flex/Flex.js");
var import_InputGroup_css = require("../../layouts/input-group/InputGroup.sty.js");
var import_InputGroup = require("../../layouts/input-group/InputGroup.js");
var import_field_css = require("../../styles/field.sty.js");
var import_textStyle_css = require("../../styles/textStyle.sty.js");
var import_DateTimePicker_css = require("../date-time-picker/DateTimePicker.sty.js");
var import_focusDateTimeInput = require("../date-time-picker/utils/focusDateTimeInput.js");
var import_FormFieldMessageContext = require("../forms-core/contexts/FormFieldMessageContext.js");
var import_useFormFieldMessagesContext = require("../forms-core/hooks/useFormFieldMessagesContext.js");
var import_useSyncedFormControlProps = require("../forms-core/hooks/useSyncedFormControlProps.js");
var import_formfield_messages_state_reducer = require("../forms-core/state/formfield-messages-state-reducer.js");
function isInputInSelectionOrFocused(ref, preRestoreEndpoints) {
  const element = ref.current?.element;
  if (!element) {
    return false;
  }
  if (element.contains(document.activeElement)) {
    return true;
  }
  const selection = window.getSelection();
  if (selection && selection.rangeCount > 0 && !selection.isCollapsed) {
    if (selection.getRangeAt(0).intersectsNode(element)) {
      return true;
    }
  }
  return !!preRestoreEndpoints && (element.contains(preRestoreEndpoints.anchorNode) || element.contains(preRestoreEndpoints.focusNode));
}
function focusExpressionInputAtEnd(ref) {
  const expressionElement = ref.current?.expressionInputRef;
  if (!expressionElement) {
    return;
  }
  expressionElement.focus();
  const selection = window.getSelection();
  if (!selection) {
    return;
  }
  const range = document.createRange();
  range.selectNodeContents(expressionElement);
  range.collapse(false);
  selection.removeAllRanges();
  selection.addRange(range);
}
function getOffsetWithinElement(element, container, offset) {
  const range = document.createRange();
  range.selectNodeContents(element);
  range.setEnd(container, offset);
  return range.toString().length;
}
const DateTimeRangePickerRoot = (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    id: propId,
    disabled: propDisabled,
    readOnly = false,
    name,
    style: consumerStyle,
    className: consumerClassName,
    required: propRequired,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "aria-disabled": ariaDisabledProp,
    placeholder: placeholderProp,
    onBlur,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const ariaProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(props);
  const COMPONENT_NAME = "DateTimeRangePicker";
  const rootRef = (0, import_react.useRef)(null);
  const touchedContainerRef = (0, import_react.useRef)(null);
  const dateTimeToRef = (0, import_react.useRef)(null);
  const dateRangeSelectorRef = (0, import_react.useRef)(null);
  const [hasInputContent, setHasInputContent] = (0, import_react.useState)(false);
  const setValue = (0, import_useTimeframeStore.useDateTimeRangePickerStore)(
    (state) => state.setDateTimeRangePickerTimeframe
  );
  const value = (0, import_useTimeframeStore.useDateTimeRangePickerStore)(
    (state) => state.dateTimeRangePickerTimeframe
  );
  const dateTimeFromRef = (0, import_useTimeframeStore.useDateTimeRangePickerStore)(
    (state) => state.dateTimeFromRef
  );
  const precision = (0, import_useTimeframeStore.useDateTimeRangePickerStore)((state) => state.precision);
  const min = (0, import_useTimeframeStore.useDateTimeRangePickerStore)((state) => state.min);
  const max = (0, import_useTimeframeStore.useDateTimeRangePickerStore)((state) => state.max);
  const formatString = (0, import_useTimeframeStore.useDateTimeRangePickerStore)(
    (state) => state.formatString
  );
  const { disabled, required, id, ariaDisabled } = (0, import_useSyncedFormControlProps.useSyncedFormControlProps)({
    propDisabled,
    propRequired,
    propId,
    propAriaDisabled: ariaDisabledProp
  });
  const disabledOrAriaDisabled = disabled || ariaDisabled || false;
  const { placeholder, onPasteCapture, onUndoPasteKeyDown } = (0, import_useDateTimeRangePicker.useDateTimeRangePicker)({
    placeholder: placeholderProp,
    readOnly,
    dateTimeFromRef,
    dateTimeToRef
  });
  const intl = (0, import_react_intl.useIntl)();
  const {
    state: { showError },
    dispatch: formMessageDispatch
  } = (0, import_useFormFieldMessagesContext.useFormFieldMessagesContext)();
  const lastValidationErrorRef = (0, import_react.useRef)();
  (0, import_react.useImperativeHandle)(
    forwardedRef,
    () => ({
      element: rootRef.current,
      inputRef: {
        from: dateTimeFromRef.current?.inputRef,
        to: dateTimeToRef.current?.inputRef
      },
      open: () => dateRangeSelectorRef.current?.open(),
      close: () => dateRangeSelectorRef.current?.close(),
      focus: () => dateTimeFromRef.current?.focus?.(),
      getCurrentValue: () => ({
        from: value?.from ?? null,
        to: value?.to ?? null
      }),
      validate() {
        lastValidationErrorRef.current = (0, import_validation.validate)(value, min, max, {
          formMessageDispatch,
          formerError: lastValidationErrorRef.current,
          precision,
          showError: true,
          checkOrder: true,
          ref: rootRef,
          expectedFormat: formatString,
          valueMode: "dateTimePicker",
          required
        });
        formMessageDispatch?.({
          type: import_formfield_messages_state_reducer.FormFieldMessagesAction.SETISTOUCHED,
          payload: { isTouched: true }
        });
      }
    }),
    [
      value,
      min,
      max,
      formMessageDispatch,
      precision,
      formatString,
      required,
      dateTimeFromRef
    ]
  );
  const { focusClassName } = (0, import_useFocusRing.useFocusRing)({
    focusWithin: true,
    ignoreModality: true,
    isMinimal: false,
    disabled,
    color: showError ? "critical" : "neutral"
  });
  const dateTimeFromLabel = intl.formatMessage({
    defaultMessage: "DateTime from",
    id: "fMy6q/KKOqbvCA3H",
    description: "Textbox to enter a relative or absolute timeframe from value."
  });
  const dateTimeToLabel = intl.formatMessage({
    defaultMessage: "DateTime to",
    id: "uuhhc3nEZptwtV2E",
    description: "Textbox to enter a relative or absolute timeframe to value."
  });
  const clearButtonLabel = intl.formatMessage({
    defaultMessage: "Clear date and time range",
    id: "zX4r/Y8gJL1C8+5e",
    description: "Aria label for the DateTimeRangePicker clear button."
  });
  const formFieldMessageContextValue = (0, import_react.useMemo)(
    () => ({
      state: {
        customMessages: [],
        formControlMessages: [],
        showError
      }
    }),
    [showError]
  );
  const { onMouseDown, onKeyDown: handleSelectionKeyDown } = (0, import_useSelectionInteractionHandlers.useSelectionInteractionHandlers)(
    touchedContainerRef,
    (0, import_react.useCallback)(() => {
      (0, import_focusDateTimeInput.focusDateTimeInput)(null, dateTimeFromRef, dateTimeToRef);
    }, [dateTimeFromRef]),
    dateTimeFromRef,
    dateTimeToRef
  );
  const onKeyDown = (0, import_react.useCallback)(
    (event) => {
      if (disabled || readOnly) {
        handleSelectionKeyDown(event);
        return;
      }
      const preRestoreEndpoints = (0, import_selection_helpers.peekPreRestoreEndpoints)(
        event.currentTarget
      );
      const fromInputIsIso = dateTimeFromRef.current?.element?.getAttribute("data-valuetype") === "iso8601";
      const toInputIsIso = dateTimeToRef.current?.element?.getAttribute("data-valuetype") === "iso8601";
      const bothInSelection = isInputInSelectionOrFocused(dateTimeFromRef, preRestoreEndpoints) && isInputInSelectionOrFocused(dateTimeToRef, preRestoreEndpoints);
      if (bothInSelection && (0, import_expression_trigger_helpers.isExpressionTriggerKey)(event)) {
        setValue(null);
        if (dateTimeToRef.current?.expressionInputRef) {
          dateTimeToRef.current.expressionInputRef.textContent = "";
        }
      }
      const switchFromToExpression = isInputInSelectionOrFocused(dateTimeFromRef, preRestoreEndpoints) && fromInputIsIso;
      const switchToToExpression = isInputInSelectionOrFocused(dateTimeToRef, preRestoreEndpoints) && toInputIsIso;
      if ((0, import_expression_trigger_helpers.isExpressionTriggerKey)(event) && (switchFromToExpression || switchToToExpression)) {
        const char = event.key;
        const primaryRef = switchFromToExpression ? dateTimeFromRef : dateTimeToRef;
        primaryRef.current?.setExpressionValue(char);
        requestAnimationFrame(() => {
          focusExpressionInputAtEnd(primaryRef);
        });
        event.preventDefault();
        event.stopPropagation();
        return;
      }
      if (isInputInSelectionOrFocused(dateTimeFromRef, preRestoreEndpoints) && isInputInSelectionOrFocused(dateTimeToRef, preRestoreEndpoints) && !fromInputIsIso && !toInputIsIso && event.key.length === 1 && !event.ctrlKey && !event.metaKey && !event.altKey) {
        const char = event.key;
        const fromExpressionElement = dateTimeFromRef.current?.expressionInputRef;
        const toExpressionElement = dateTimeToRef.current?.expressionInputRef;
        const selection = window.getSelection();
        if (fromExpressionElement && toExpressionElement && selection && selection.rangeCount > 0 && !selection.isCollapsed) {
          const range = selection.getRangeAt(0);
          const fromValue = fromExpressionElement.textContent ?? "";
          const toValue = toExpressionElement.textContent ?? "";
          const fromStart = fromExpressionElement.contains(range.startContainer) ? getOffsetWithinElement(
            fromExpressionElement,
            range.startContainer,
            range.startOffset
          ) : 0;
          const toEnd = toExpressionElement.contains(range.endContainer) ? getOffsetWithinElement(
            toExpressionElement,
            range.endContainer,
            range.endOffset
          ) : toValue.length;
          dateTimeFromRef.current?.setExpressionValue(
            `${fromValue.slice(0, fromStart)}${char}`
          );
          dateTimeToRef.current?.setExpressionValue(toValue.slice(toEnd));
          requestAnimationFrame(() => {
            focusExpressionInputAtEnd(dateTimeFromRef);
          });
          event.preventDefault();
          event.stopPropagation();
          return;
        }
        dateTimeFromRef.current?.setExpressionValue(char);
        requestAnimationFrame(() => {
          focusExpressionInputAtEnd(dateTimeFromRef);
        });
        event.preventDefault();
        event.stopPropagation();
        return;
      }
      handleSelectionKeyDown(event);
    },
    [
      disabled,
      readOnly,
      handleSelectionKeyDown,
      dateTimeFromRef,
      dateTimeToRef,
      setValue
    ]
  );
  const handleFromTrailingEdge = (0, import_react.useCallback)(() => {
    dateTimeToRef.current?.focus?.();
  }, []);
  const handleToLeadingEdge = (0, import_react.useCallback)(() => {
    dateTimeFromRef.current?.focusLast();
  }, [dateTimeFromRef]);
  const handleFromTrailingOverflow = (0, import_react.useCallback)(
    (overflowChar) => {
      const handlers = Array.from(
        dateTimeToRef.current?.spinButtonHandlers?.values() ?? []
      ).filter((h) => h !== null && h !== void 0);
      const firstHandler = handlers[0];
      if (!firstHandler) {
        return;
      }
      if (value?.to?.type === "expression" || value?.to === null) {
        dateTimeToRef.current?.setExpressionValue(overflowChar);
        requestAnimationFrame(() => {
          const exprEl = dateTimeToRef.current?.expressionInputRef;
          if (exprEl) {
            exprEl.focus();
            const sel = window.getSelection();
            if (sel) {
              const range = document.createRange();
              range.selectNodeContents(exprEl);
              range.collapse(false);
              sel.removeAllRanges();
              sel.addRange(range);
            }
          }
        });
        return;
      }
      if (firstHandler.type === "number") {
        firstHandler.applyText(overflowChar, true);
        firstHandler.element.current?.focus();
        firstHandler.setSelection({ start: 0, end: 0, rtl: true });
      }
    },
    [value?.to]
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Flex.Flex,
    {
      ref: rootRef,
      flexDirection: "column",
      "data-testid": dataTestId,
      style: consumerStyle,
      className: consumerClassName,
      ...ariaProps,
      ...behavioralTrackingProps,
      "data-dt-component": COMPONENT_NAME,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_FormFieldMessageContext.FormFieldMessageContext.Provider, { value: formFieldMessageContextValue, children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_InputGroup.InputGroup, { style: { gridTemplateColumns: "1fr auto" }, children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
          import_Flex.Flex,
          {
            ref: touchedContainerRef,
            className: (0, import_clsx.default)(
              import_InputGroup_css.inputGroupClassName,
              (0, import_DateTimePicker_css.dateTimePicker)({
                readOnly,
                showError,
                disabled,
                type: "datetime"
              }),
              (0, import_textStyle_css.textStyle)({ textStyle: "base" }),
              (0, import_field_css.fieldTokens)({
                color: showError ? "critical" : "neutral",
                disabled: disabledOrAriaDisabled,
                variant: "emphasized"
              }),
              focusClassName,
              styles.dateTimeRange({
                showError,
                disabled: disabledOrAriaDisabled,
                readOnly
              })
            ),
            onCopyCapture: import_selection_copy_helpers.onTimeframeCopyCapture,
            onPasteCapture,
            onBlur: (event) => {
              requestAnimationFrame(() => {
                onBlur?.(event);
              });
            },
            onInput: () => {
              setHasInputContent(
                !!dateTimeFromRef.current?.expressionInputRef?.textContent || !!dateTimeToRef.current?.expressionInputRef?.textContent
              );
            },
            tabIndex: disabled ? void 0 : 0,
            onMouseDown,
            onKeyDownCapture: (e) => {
              onUndoPasteKeyDown(e);
            },
            onKeyDown,
            onFocus: (e) => {
              if (!rootRef.current?.contains(e.relatedTarget)) {
                (0, import_focusDateTimeInput.focusDateTimeInput)(e, dateTimeFromRef, dateTimeToRef);
              }
            },
            children: [
              /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("span", { className: styles.scrollContainer, children: [
                /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  import_DateTimeInput.DateTimeInput,
                  {
                    ref: dateTimeFromRef,
                    className: styles.dateTimeInput,
                    disabled,
                    readOnly,
                    placeholder: placeholder.from,
                    name: name === void 0 ? void 0 : `${name}-from`,
                    value: value === null ? null : value.from,
                    onChange: (newValue) => {
                      if (typeof newValue !== "string") {
                        setValue({ from: newValue ?? null });
                      }
                    },
                    "aria-label": dateTimeFromLabel,
                    "aria-disabled": ariaDisabled,
                    required,
                    controlId: id,
                    onTrailingEdge: handleFromTrailingEdge,
                    onTrailingOverflow: handleFromTrailingOverflow,
                    partialTimeAnchorDate: value?.to?.absoluteDate || void 0,
                    onPasteComplete: () => {
                      dateTimeToRef.current?.focus?.();
                    }
                  }
                ),
                /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { className: styles.timeframeSeparator, children: "\u2192" }),
                /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  import_DateTimeInput.DateTimeInput,
                  {
                    ref: dateTimeToRef,
                    className: styles.dateTimeInput,
                    disabled,
                    readOnly,
                    placeholder: placeholder.to,
                    name: name === void 0 ? void 0 : `${name}-to`,
                    value: value === null ? null : value.to,
                    onChange: (newValue) => {
                      if (typeof newValue !== "string") {
                        setValue({ to: newValue ?? null });
                      }
                    },
                    "aria-label": dateTimeToLabel,
                    "aria-disabled": ariaDisabled,
                    required,
                    onLeadingEdge: handleToLeadingEdge,
                    dateOnlyPasteAnchor: "endOfDay",
                    partialTimeAnchorDate: value?.from?.absoluteDate || void 0,
                    onPasteComplete: () => {
                      requestAnimationFrame(() => {
                        dateTimeToRef.current?.focusLast();
                      });
                    }
                  }
                )
              ] }),
              !readOnly && !disabledOrAriaDisabled && (value && (value.from?.value !== "" || value.to?.value !== "") || hasInputContent) && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_Button.Button,
                {
                  onFocus: () => {
                    formMessageDispatch?.({
                      type: import_formfield_messages_state_reducer.FormFieldMessagesAction.SETISTOUCHED,
                      payload: { isTouched: true }
                    });
                  },
                  onClick: () => {
                    setValue(null);
                    if (dateTimeFromRef.current?.expressionInputRef) {
                      dateTimeFromRef.current.expressionInputRef.textContent = "";
                    }
                    if (dateTimeToRef.current?.expressionInputRef) {
                      dateTimeToRef.current.expressionInputRef.textContent = "";
                    }
                    setHasInputContent(false);
                    requestAnimationFrame(() => {
                      dateTimeFromRef.current?.expressionInputRef?.focus();
                    });
                  },
                  onKeyDown: (e) => {
                    if (e.key === "Enter") {
                      e.stopPropagation();
                    }
                  },
                  className: import_DateTimePicker_css.clearIconButton,
                  color: showError ? "critical" : "neutral",
                  size: "condensed",
                  "aria-label": clearButtonLabel,
                  children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.XmarkIcon, {})
                }
              )
            ]
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_DateRangeSelector.DateRangeSelector,
          {
            ref: dateRangeSelectorRef,
            disabled,
            "aria-disabled": ariaDisabled,
            readOnly,
            error: showError,
            "data-dtrum-mask": dataDtrumMask,
            "data-dtrum-allow": dataDtrumAllow
          }
        )
      ] }) })
    }
  );
});
