// @dynatrace/strato-components/forms/date-time-range-picker/helpers.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var helpers_exports = {};
__export(helpers_exports, {
  captureUndoSnapshot: () => captureUndoSnapshot,
  isContentModifyingKey: () => isContentModifyingKey,
  normalizeISOInput: () => normalizeISOInput,
  parseDateOnlyToISO: () => parseDateOnlyToISO,
  parseTimeOnlyToISO: () => parseTimeOnlyToISO,
  partialDateTimeToISO: () => partialDateTimeToISO,
  tryParsePartialDate: () => tryParsePartialDate,
  tryParsePartialTime: () => tryParsePartialTime
});
module.exports = __toCommonJS(helpers_exports);
var import_tz = require("@date-fns/tz");
var import_date_fns = require("date-fns");
var import_user_preferences = require("@dynatrace-sdk/user-preferences");
var import_helpers = require("../forms-core/time/helpers.js");
const DATE_ONLY_REGEX = /^(\d{4})-(0[1-9]|1[0-2])-(0[1-9]|[12]\d|3[01])[Zz]?$/;
const TIME_ONLY_REGEX = /^([01]\d|2[0-3]):([0-5]\d)(?::([0-5]\d)(?:\.(\d{1,3}))?)?([Zz]|[+-](?:[01]\d|2[0-3]):[0-5]\d)?$/;
const FULL_ISO_DATETIME_REGEX = /^(\d{4})-(0[1-9]|1[0-2])-(0[1-9]|[12]\d|3[01])[Tt]([01]\d|2[0-3])(?::([0-5]\d)(?::([0-5]\d)(?:\.(\d{1,3}))?)?)?([Zz]|[+-]\d{2}:?\d{2})?$/;
function normalizeISOInput(input) {
  const match = FULL_ISO_DATETIME_REGEX.exec(input);
  if (!match) {
    return input;
  }
  let normalized = input.replace(/^(\d{4}-\d{2}-\d{2})[Tt]/, "$1T");
  const tzSuffix = match[8];
  if (tzSuffix) {
    if (tzSuffix === "z") {
      normalized = normalized.slice(0, -1) + "Z";
    }
  } else {
    const year = Number(match[1]);
    const month = Number(match[2]) - 1;
    const day = Number(match[3]);
    const hours = Number(match[4]);
    const minutes = match[5] !== void 0 ? Number(match[5]) : 0;
    const seconds = match[6] !== void 0 ? Number(match[6]) : 0;
    const milliseconds = match[7] !== void 0 ? Number(match[7].padEnd(3, "0")) : 0;
    const local = import_tz.TZDateMini.tz(
      (0, import_user_preferences.getTimezone)(),
      year,
      month,
      day,
      hours,
      minutes,
      seconds,
      milliseconds
    );
    if (!isNaN(local.getTime())) {
      if (local.getMonth() !== month || local.getDate() !== day) {
        return input;
      }
      return local.toISOString();
    }
    normalized += "Z";
  }
  return normalized;
}
function parseTimeOnlyToISO(input) {
  const match = TIME_ONLY_REGEX.exec(input);
  if (!match) {
    return null;
  }
  const hours = Number(match[1]);
  const minutes = Number(match[2]);
  const seconds = match[3] !== void 0 ? Number(match[3]) : 0;
  const milliseconds = match[4] !== void 0 ? Number(match[4].padEnd(3, "0")) : 0;
  const timezoneSuffix = match[5];
  const now = import_tz.TZDateMini.tz((0, import_user_preferences.getTimezone)());
  const year = now.getFullYear();
  const month = now.getMonth();
  const day = now.getDate();
  if (timezoneSuffix) {
    const ISOdate = `${year}-${String(month + 1).padStart(2, "0")}-${String(
      day
    ).padStart(2, "0")}`;
    const ISOtime = `T${match[1]}:${match[2]}:${String(seconds).padStart(
      2,
      "0"
    )}.${String(milliseconds).padStart(3, "0")}${timezoneSuffix.toUpperCase()}`;
    const fullISO = ISOdate + ISOtime;
    const parsedDate = new Date(fullISO);
    return isNaN(parsedDate.getTime()) ? null : parsedDate.toISOString();
  }
  const local = import_tz.TZDateMini.tz(
    (0, import_user_preferences.getTimezone)(),
    year,
    month,
    day,
    hours,
    minutes,
    seconds,
    milliseconds
  );
  return isNaN(local.getTime()) ? null : local.toISOString();
}
function parseDateOnlyToISO(input, anchor = "startOfDay") {
  const match = DATE_ONLY_REGEX.exec(input);
  if (!match) {
    return null;
  }
  const date = import_tz.TZDateMini.tz(
    (0, import_user_preferences.getTimezone)(),
    Number(match[1]),
    Number(match[2]) - 1,
    // 0-indexed month
    Number(match[3])
  );
  if (date.getMonth() !== Number(match[2]) - 1) {
    return null;
  }
  const anchored = anchor === "endOfDay" ? (0, import_date_fns.endOfDay)(date) : (0, import_date_fns.startOfDay)(date);
  return anchored.toISOString();
}
const captureUndoSnapshot = (spinButtonHandlers, expressionRef, value, valueType) => {
  const spinButtonTexts = /* @__PURE__ */ new Map();
  spinButtonHandlers.forEach((handler, type) => {
    spinButtonTexts.set(type, handler?.element.current?.textContent ?? "");
  });
  return {
    value,
    valueType,
    spinButtonTexts,
    expressionText: expressionRef?.textContent ?? ""
  };
};
function isContentModifyingKey(event) {
  if (event.ctrlKey || event.metaKey) {
    return false;
  }
  if (event.key === "ArrowUp" || event.key === "ArrowDown" || event.key === "Backspace" || event.key === "Delete") {
    return true;
  }
  if (event.key.length === 1) {
    return true;
  }
  return false;
}
function isDateFieldPart(part) {
  return part.type === "day" || part.type === "month" || part.type === "year";
}
function getRegionalDateFormat() {
  const parts = new Intl.DateTimeFormat((0, import_helpers.getSafeRegionalFormat)(), {
    year: "numeric",
    month: "2-digit",
    day: "2-digit"
  }).formatToParts(/* @__PURE__ */ new Date());
  const delimiter = parts.find((p) => p.type === "literal")?.value ?? "/";
  const fieldOrder = parts.filter(isDateFieldPart).map((p) => p.type);
  return { delimiter, fieldOrder };
}
const PLACEHOLDER_TOKEN_PATTERN = "[A-Za-z]{1,4}";
function orPlaceholder(numericPattern) {
  return `${numericPattern}|${PLACEHOLDER_TOKEN_PATTERN}`;
}
function isPlaceholderToken(value) {
  return /^[A-Za-z]+$/.test(value);
}
function tryParsePartialDate(input) {
  const { delimiter, fieldOrder } = getRegionalDateFormat();
  const escapedDelimiter = delimiter.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  const regex = new RegExp(
    `(?<=^|[\\s,])(${orPlaceholder(
      "\\d{1,4}"
    )})${escapedDelimiter}(${orPlaceholder(
      "\\d{1,4}"
    )})${escapedDelimiter}?(${orPlaceholder("\\d{1,4}")})?(?=[\\s,]|$)`
  );
  const match = regex.exec(input);
  if (!match) {
    return null;
  }
  const rawValues = [match[1], match[2], match[3]].filter(
    (value) => value !== void 0
  );
  const looksLikeYear = (value) => value.length === 4;
  const hasDisplacedYear = rawValues.some(
    (value, index) => looksLikeYear(value) && fieldOrder[index] !== "year"
  );
  const isYearFirstLocale = fieldOrder[0] === "year";
  const hasNoExplicitYear = !rawValues.some(looksLikeYear);
  const fields = hasDisplacedYear || isYearFirstLocale && hasNoExplicitYear ? fieldOrder.slice(fieldOrder.length - rawValues.length) : fieldOrder.slice(0, rawValues.length);
  const maxDigits = {
    year: 4,
    day: 2,
    month: 2
  };
  const result = {};
  for (let i = 0; i < fields.length; i++) {
    const raw = rawValues[i];
    const field = fields[i];
    if (isPlaceholderToken(raw)) {
      continue;
    }
    if (raw.length > maxDigits[field]) {
      return null;
    }
    const value = parseInt(raw);
    if (field === "month" && (value < 1 || value > 12)) {
      return null;
    }
    if (field === "day" && (value < 1 || value > 31)) {
      return null;
    }
    result[field] = value;
  }
  return result;
}
function getRegionalTimeFormat() {
  const parts = new Intl.DateTimeFormat((0, import_helpers.getSafeRegionalFormat)(), {
    hour: "numeric",
    minute: "2-digit"
  }).formatToParts(/* @__PURE__ */ new Date());
  const hasDayPeriod = parts.some((p) => p.type === "dayPeriod");
  const dayPeriodValues = (0, import_helpers.getDayPeriodValues)();
  return { is12Hour: hasDayPeriod, dayPeriodValues };
}
function parseTimeField(raw) {
  if (raw === void 0 || isPlaceholderToken(raw)) {
    return void 0;
  }
  return parseInt(raw);
}
function tryParsePartialTime(input) {
  const { is12Hour, dayPeriodValues } = getRegionalTimeFormat();
  const escapedPeriods = dayPeriodValues.map(
    (value) => value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")
  );
  const dayPeriodPattern = is12Hour ? `(?:\\s*(${escapedPeriods.join("|")}|[A-Za-z]+))?` : "";
  const hourPattern = orPlaceholder("[01]?\\d|2[0-3]");
  const minuteOrSecondPattern = orPlaceholder("[0-5]\\d");
  const millisecondPattern = orPlaceholder("\\d{1,3}");
  const regex = new RegExp(
    `(?<=^|[\\s,])(${hourPattern}):(${minuteOrSecondPattern})(?::(${minuteOrSecondPattern})(?:[.](${millisecondPattern}))?)?${dayPeriodPattern}(?=[\\s,]|$)`,
    "i"
  );
  const match = regex.exec(input);
  if (match) {
    const result = {};
    result.hour = parseTimeField(match[1]);
    result.minute = parseTimeField(match[2]);
    result.second = parseTimeField(match[3]);
    result.millisecond = parseTimeField(match[4]);
    if (is12Hour && match[5] !== void 0) {
      const matched = match[5].toLowerCase();
      result.dayPeriod = dayPeriodValues.find(
        (p) => p.toLowerCase() === matched
      );
    }
    return result;
  }
  const millisecondsRegex = /(?<=^|[\s,])([0-5]\d):([0-5]\d)\.(\d{1,3})(?=[\s,]|$)/;
  const millisecondsMatch = millisecondsRegex.exec(input);
  if (millisecondsMatch) {
    return {
      minute: parseInt(millisecondsMatch[1]),
      second: parseInt(millisecondsMatch[2]),
      millisecond: parseInt(millisecondsMatch[3])
    };
  }
  return null;
}
function convertTo24Hour(hour, dayPeriod) {
  const { dayPeriodValues } = getRegionalTimeFormat();
  const isPM = dayPeriod.toLowerCase() === dayPeriodValues[1].toLowerCase();
  const isAM = dayPeriod.toLowerCase() === dayPeriodValues[0].toLowerCase();
  if (isPM && hour < 12) {
    return hour + 12;
  }
  if (isAM && hour === 12) {
    return 0;
  }
  return hour;
}
function buildFullDateTimeISO(year, month, day, hour, minute, second, millisecond) {
  const d = import_tz.TZDateMini.tz(
    (0, import_user_preferences.getTimezone)(),
    year,
    month - 1,
    day,
    hour,
    minute,
    second ?? 0,
    millisecond ?? 0
  );
  return isNaN(d.getTime()) ? null : { iso: d.toISOString() };
}
function buildDateOnlyISO(year, month, day) {
  const validationDate = import_tz.TZDateMini.tz((0, import_user_preferences.getTimezone)(), year, month - 1, day);
  if (isNaN(validationDate.getTime()) || validationDate.getFullYear() !== year || validationDate.getMonth() !== month - 1 || validationDate.getDate() !== day) {
    return null;
  }
  const yearString = String(year).padStart(4, "0");
  const monthString = String(month).padStart(2, "0");
  const dayString = String(day).padStart(2, "0");
  return { iso: `${yearString}-${monthString}-${dayString}` };
}
function buildTimeOnlyISO(hour, minute, second, millisecond) {
  const hourString = String(hour).padStart(2, "0");
  const minuteString = String(minute).padStart(2, "0");
  let isoTime = `${hourString}:${minuteString}`;
  if (second !== void 0) {
    const secondString = String(second).padStart(2, "0");
    isoTime += `:${secondString}`;
    if (millisecond !== void 0) {
      const millisecondString = String(millisecond).padStart(3, "0");
      isoTime += `.${millisecondString}`;
    }
  }
  return { iso: isoTime };
}
function collectPartsToUpdate(date, time) {
  const partsToUpdate = /* @__PURE__ */ new Set();
  const fieldMapping = [
    [date?.day, "day"],
    [date?.month, "month"],
    [date?.year, "year"],
    [time?.hour, "hour"],
    [time?.minute, "minute"],
    [time?.second, "second"],
    [time?.millisecond, "fractionalSecond"],
    [time?.dayPeriod, "dayPeriod"]
  ];
  for (const [value, part] of fieldMapping) {
    if (value !== void 0) {
      partsToUpdate.add(part);
    }
  }
  return partsToUpdate;
}
function partialDateTimeToISO(date, time) {
  if (!date && !time) {
    return null;
  }
  const year = date?.year;
  const month = date?.month;
  const day = date?.day;
  const minute = time?.minute;
  const second = time?.second;
  const millisecond = time?.millisecond;
  let hour24 = time?.hour;
  if (hour24 !== void 0 && time?.dayPeriod) {
    hour24 = convertTo24Hour(hour24, time.dayPeriod);
  }
  if (year !== void 0 && month !== void 0 && day !== void 0) {
    if (hour24 !== void 0 && minute !== void 0) {
      return buildFullDateTimeISO(
        year,
        month,
        day,
        hour24,
        minute,
        second,
        millisecond
      );
    }
    if (time === null) {
      return buildDateOnlyISO(year, month, day);
    }
  }
  if (date === null && hour24 !== void 0 && minute !== void 0) {
    return buildTimeOnlyISO(hour24, minute, second, millisecond);
  }
  const now = import_tz.TZDateMini.tz((0, import_user_preferences.getTimezone)());
  const fallback = import_tz.TZDateMini.tz(
    (0, import_user_preferences.getTimezone)(),
    year ?? now.getFullYear(),
    (month ?? now.getMonth() + 1) - 1,
    day ?? now.getDate(),
    hour24 ?? now.getHours(),
    minute ?? now.getMinutes(),
    second ?? 0,
    millisecond ?? 0
  );
  if (isNaN(fallback.getTime())) {
    return null;
  }
  const partsToUpdate = collectPartsToUpdate(date, time);
  if (partsToUpdate.size === 0) {
    return null;
  }
  return { iso: fallback.toISOString(), partsToUpdate };
}
