// @dynatrace/strato-components/forms/date-time-range-picker/selection-helpers.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var selection_helpers_exports = {};
__export(selection_helpers_exports, {
  buildSpinButtonSnapshot: () => buildSpinButtonSnapshot,
  checkIfValueTypeAlignment: () => checkIfValueTypeAlignment,
  collapseMultiElementSelection: () => collapseMultiElementSelection,
  consumePreRestoreEndpoints: () => consumePreRestoreEndpoints,
  deleteCrossExpressionPartial: () => deleteCrossExpressionPartial,
  deletePartialExpression: () => deletePartialExpression,
  extendSelection: () => extendSelection,
  extendSelectionToAll: () => extendSelectionToAll,
  findExpressionFallbacks: () => findExpressionFallbacks,
  findFirstExpressionFocusTarget: () => findFirstExpressionFocusTarget,
  focusExpression: () => focusExpression,
  getCrossExpressionElements: () => getCrossExpressionElements,
  getExpressionCursorOffset: () => getExpressionCursorOffset,
  getFirstSnapshotHandler: () => getFirstSnapshotHandler,
  getLastSnapshotHandler: () => getLastSnapshotHandler,
  getPartialExpressionElement: () => getPartialExpressionElement,
  peekPreRestoreEndpoints: () => peekPreRestoreEndpoints,
  planBulkClear: () => planBulkClear,
  removeContentEditable: () => removeContentEditable,
  restoreExpressionCursorOffset: () => restoreExpressionCursorOffset,
  restoreFocusAfterBulkClear: () => restoreFocusAfterBulkClear,
  restoreSpinButtonSnapshot: () => restoreSpinButtonSnapshot
});
module.exports = __toCommonJS(selection_helpers_exports);
var styles = __toESM(require("./DateTimeInput.sty.js"));
var import_sdk_logger = require("../../core/sdk/sdk-logger.js");
function warnSelectionFailed(message, error) {
  (0, import_sdk_logger.getStratoLogger)("DateTimeRangePicker").warn(message, error);
}
function checkIfValueTypeAlignment(textNode) {
  const closestDateTimeInput = textNode.parentElement?.closest(
    `.${styles.dateTimeInput}`
  );
  const currentValueType = closestDateTimeInput?.getAttribute("data-valuetype");
  const closestValueTypeElement = textNode.parentElement?.closest("[data-valuetype]");
  if (closestDateTimeInput === closestValueTypeElement) {
    return true;
  }
  const nodeValueType = closestValueTypeElement?.getAttribute("data-valuetype");
  return nodeValueType === currentValueType;
}
function previousTextNode(root, node) {
  const walker = document.createTreeWalker(root, NodeFilter.SHOW_TEXT, null);
  let current = walker.lastChild();
  let found = false;
  while (current) {
    if (found) {
      if (checkIfValueTypeAlignment(current)) {
        return current;
      }
    }
    if (current === node) {
      found = true;
    }
    current = walker.previousNode();
  }
  return null;
}
function nextTextNode(root, node) {
  const walker = document.createTreeWalker(root, NodeFilter.SHOW_TEXT, null);
  let current = walker.nextNode();
  let found = false;
  while (current) {
    if (found) {
      if (checkIfValueTypeAlignment(current)) {
        return current;
      }
    }
    if (current === node) {
      found = true;
    }
    current = walker.nextNode();
  }
  return null;
}
function findNextCharacterPosition(root, node, offset) {
  if (node.nodeType === Node.TEXT_NODE && node.nodeValue && offset < node.nodeValue.length && checkIfValueTypeAlignment(node)) {
    return { node, offset: offset + 1 };
  }
  if (node.nodeType !== Node.TEXT_NODE) {
    const childAtOffset = node.childNodes[offset];
    if (childAtOffset) {
      const walker = document.createTreeWalker(
        childAtOffset,
        NodeFilter.SHOW_TEXT,
        null
      );
      const firstText = walker.firstChild() ?? (childAtOffset.nodeType === Node.TEXT_NODE ? childAtOffset : null);
      if (firstText && (checkIfValueTypeAlignment(firstText) || getClosestContentEditable(firstText) !== null)) {
        return { node: firstText, offset: 1 };
      }
    }
    if (offset > 0) {
      const previousChild = node.childNodes[offset - 1];
      if (previousChild) {
        const walker = document.createTreeWalker(
          previousChild,
          NodeFilter.SHOW_TEXT,
          null
        );
        const lastText = walker.lastChild() ?? (previousChild.nodeType === Node.TEXT_NODE ? previousChild : null);
        if (lastText) {
          let nextFromBoundary = nextTextNode(root, lastText);
          while (nextFromBoundary !== null && getClosestContentEditable(nextFromBoundary) === null) {
            nextFromBoundary = nextTextNode(root, nextFromBoundary);
          }
          if (nextFromBoundary) {
            return { node: nextFromBoundary, offset: 1 };
          }
        }
      }
    }
  }
  let next = nextTextNode(root, node);
  if (isExpressionNode(node) && next && getClosestContentEditable(next) === null) {
    const nextEditableNode = nextTextNode(root, next);
    if (isExpressionNode(nextEditableNode)) {
      next = nextEditableNode;
    }
  }
  if (!next) {
    return null;
  }
  return { node: next, offset: 1 };
}
function findPreviousCharacterPosition(root, node, offset) {
  if (node.nodeType === Node.TEXT_NODE && offset > 0 && checkIfValueTypeAlignment(node)) {
    return { node, offset: offset - 1 };
  }
  if (node.nodeType !== Node.TEXT_NODE && offset > 0) {
    const childAtOffset = node.childNodes[offset - 1];
    if (childAtOffset) {
      const walker = document.createTreeWalker(
        childAtOffset,
        NodeFilter.SHOW_TEXT,
        null
      );
      const lastText = walker.lastChild() ?? (childAtOffset.nodeType === Node.TEXT_NODE ? childAtOffset : null);
      if (lastText?.nodeValue && (checkIfValueTypeAlignment(lastText) || getClosestContentEditable(lastText) !== null)) {
        return { node: lastText, offset: lastText.nodeValue.length - 1 };
      }
    }
  }
  let prev = previousTextNode(root, node);
  if (isExpressionNode(node) && prev && getClosestContentEditable(prev) === null) {
    const previousEditableNode = previousTextNode(root, prev);
    if (isExpressionNode(previousEditableNode)) {
      prev = previousEditableNode;
    }
  }
  if (!prev || prev?.nodeValue === null) {
    return null;
  }
  return { node: prev, offset: prev.nodeValue.length - 1 };
}
function isEmptyPlaceholderSpinButton(node) {
  return node instanceof Element && node.getAttribute("role") === "spinbutton" && node.childNodes.length === 0;
}
function createSelectionPositionWalker(root) {
  return document.createTreeWalker(
    root,
    NodeFilter.SHOW_TEXT | NodeFilter.SHOW_ELEMENT,
    {
      acceptNode: (node) => {
        if (node.nodeType === Node.TEXT_NODE) {
          return NodeFilter.FILTER_ACCEPT;
        }
        return isEmptyPlaceholderSpinButton(node) ? NodeFilter.FILTER_ACCEPT : NodeFilter.FILTER_SKIP;
      }
    }
  );
}
function findFirstCharacterPosition(root) {
  const walker = createSelectionPositionWalker(root);
  let current = walker.firstChild();
  while (current !== null) {
    if (checkIfValueTypeAlignment(current)) {
      return { node: current, offset: 0 };
    }
    current = walker.nextNode();
  }
  return null;
}
function findLastCharacterPosition(root) {
  const walker = createSelectionPositionWalker(root);
  let current = walker.lastChild();
  while (current !== null) {
    if (checkIfValueTypeAlignment(current)) {
      const offset = current.nodeType === Node.TEXT_NODE ? current.nodeValue?.length ?? 0 : 0;
      return { node: current, offset };
    }
    current = walker.previousNode();
  }
  return null;
}
function compareSelectionPoints(firstNode, firstOffset, secondNode, secondOffset) {
  if (firstNode === secondNode) {
    return firstOffset - secondOffset;
  }
  if (firstNode.getRootNode() !== secondNode.getRootNode()) {
    return 0;
  }
  const firstRange = document.createRange();
  firstRange.setStart(firstNode, firstOffset);
  firstRange.collapse(true);
  const secondRange = document.createRange();
  secondRange.setStart(secondNode, secondOffset);
  secondRange.collapse(true);
  try {
    return firstRange.compareBoundaryPoints(Range.START_TO_START, secondRange);
  } catch {
    const pos = firstNode.compareDocumentPosition(secondNode);
    if (pos & Node.DOCUMENT_POSITION_FOLLOWING) {
      return -1;
    }
    if (pos & Node.DOCUMENT_POSITION_PRECEDING) {
      return 1;
    }
    return 0;
  }
}
function selectionOverlapsSpinButton(selectionStartNode, selectionStartOffset, selectionEndNode, selectionEndOffset, spinButton) {
  const spinButtonStart = findFirstCharacterPosition(spinButton);
  const spinButtonEnd = findLastCharacterPosition(spinButton);
  if (!spinButtonStart || !spinButtonEnd) {
    return false;
  }
  return compareSelectionPoints(
    selectionStartNode,
    selectionStartOffset,
    spinButtonEnd.node,
    spinButtonEnd.offset
  ) < 0 && compareSelectionPoints(
    selectionEndNode,
    selectionEndOffset,
    spinButtonStart.node,
    spinButtonStart.offset
  ) > 0;
}
function restoreContentEditableAttributes(root) {
  for (const editable of root.querySelectorAll(
    "[data-wascontenteditable]"
  )) {
    const previousValue = editable.getAttribute("data-wascontenteditable");
    if (previousValue !== null) {
      editable.setAttribute("contenteditable", previousValue);
    }
    editable.removeAttribute("data-wascontenteditable");
    const wasTabIndex = editable.getAttribute("data-wastabindex");
    if (wasTabIndex !== null) {
      editable.setAttribute("tabindex", wasTabIndex);
      editable.removeAttribute("data-wastabindex");
    }
  }
  for (const el of root.querySelectorAll("[data-drag-selected]")) {
    el.removeAttribute("data-drag-selected");
  }
}
function markDragSelectedElements(rootElement, range, { includeTextboxes = false } = {}) {
  for (const el of rootElement.querySelectorAll(
    "[data-drag-selected]"
  )) {
    el.removeAttribute("data-drag-selected");
  }
  for (const group of rootElement.querySelectorAll(
    '[data-valuetype="iso8601"]'
  )) {
    for (const child of group.children) {
      if (range.intersectsNode(child)) {
        child.setAttribute("data-drag-selected", "");
      }
    }
  }
  if (includeTextboxes) {
    for (const textbox of rootElement.querySelectorAll('[role="textbox"]')) {
      if (range.intersectsNode(textbox)) {
        textbox.setAttribute("data-drag-selected", "");
      }
    }
  }
}
const pendingRestoreSignals = /* @__PURE__ */ new WeakMap();
const lastPreRestoreEndpoints = /* @__PURE__ */ new WeakMap();
function consumePreRestoreEndpoints(root) {
  const stash = lastPreRestoreEndpoints.get(root);
  lastPreRestoreEndpoints.delete(root);
  return stash;
}
function shouldBypassRestore(e) {
  if (["Meta", "Control", "Alt", "Shift"].includes(e.key)) {
    return true;
  }
  if ((e.metaKey || e.ctrlKey) && e.key.toLowerCase() === "c") {
    return true;
  }
  return e.shiftKey && ["ArrowLeft", "ArrowRight", "ArrowUp", "ArrowDown", "Home", "End"].includes(
    e.key
  );
}
function tryCollapseArrow(e, root) {
  if (e.shiftKey || e.metaKey || e.ctrlKey || e.altKey || e.key !== "ArrowLeft" && e.key !== "ArrowRight") {
    return false;
  }
  const collapsed = collapseMultiElementSelection(
    root,
    e.key === "ArrowLeft" ? "left" : "right"
  );
  if (collapsed) {
    e.preventDefault();
    e.stopPropagation();
  }
  return collapsed;
}
function peekPreRestoreEndpoints(root) {
  return lastPreRestoreEndpoints.get(root);
}
function deferContentEditableRestore(root) {
  pendingRestoreSignals.get(root)?.abort();
  const restoreSignal = new AbortController();
  pendingRestoreSignals.set(root, restoreSignal);
  function handler(e) {
    if (e instanceof KeyboardEvent) {
      if (shouldBypassRestore(e)) {
        return;
      }
      if (tryCollapseArrow(e, root)) {
        return;
      }
    }
    const selection = globalThis.getSelection();
    const anchorNode = selection?.anchorNode ?? null;
    const anchorOffset = selection?.anchorOffset ?? 0;
    const focusNode = selection?.focusNode ?? null;
    const focusOffset = selection?.focusOffset ?? 0;
    const hasMultiElementSelection = selection !== null && anchorNode !== null && focusNode !== null && anchorNode !== focusNode && selection.toString().length > 0;
    if (hasMultiElementSelection && anchorNode && focusNode && e instanceof KeyboardEvent) {
      lastPreRestoreEndpoints.set(root, {
        anchorNode,
        anchorOffset,
        focusNode,
        focusOffset
      });
    }
    restoreContentEditableAttributes(root);
    pendingRestoreSignals.delete(root);
    restoreSignal.abort();
    if (hasMultiElementSelection && anchorNode && focusNode) {
      try {
        const anchorIsEmpty = isEmptyPlaceholderSpinButton(anchorNode);
        const focusIsEmpty = isEmptyPlaceholderSpinButton(focusNode);
        if (anchorIsEmpty && !focusIsEmpty && focusNode.nodeType === Node.TEXT_NODE) {
          selection?.setBaseAndExtent(focusNode, 0, focusNode, focusOffset);
        } else if (focusIsEmpty && !anchorIsEmpty && anchorNode.nodeType === Node.TEXT_NODE) {
          const anchorTextLength = anchorNode.nodeValue?.length ?? anchorOffset;
          selection?.setBaseAndExtent(
            anchorNode,
            anchorOffset,
            anchorNode,
            anchorTextLength
          );
        } else if (!anchorIsEmpty && !focusIsEmpty) {
          selection?.setBaseAndExtent(
            anchorNode,
            anchorOffset,
            focusNode,
            focusOffset
          );
        }
      } catch (error) {
        warnSelectionFailed(
          "setBaseAndExtent failed when restoring selection after CE re-attach",
          error
        );
      }
    }
  }
  document.addEventListener("mousedown", handler, {
    signal: restoreSignal.signal,
    capture: true
  });
  document.addEventListener("keydown", handler, {
    signal: restoreSignal.signal,
    capture: true
  });
}
function removeContentEditable(rootElement, removeEventTarget, eventType, customRemovalFn, initialAnchorSpinButton) {
  rootElement.setAttribute("data-restoring-selection", "");
  const contentEditables = rootElement.querySelectorAll("[contenteditable]");
  for (const editable of contentEditables) {
    const previousValue = editable.getAttribute("contenteditable") ?? void 0;
    editable.removeAttribute("contenteditable");
    editable.dataset.wascontenteditable = previousValue;
    if (editable.hasAttribute("tabindex")) {
      editable.dataset.wastabindex = editable.getAttribute("tabindex") ?? "";
      editable.removeAttribute("tabindex");
    }
  }
  const signal = new AbortController();
  if (initialAnchorSpinButton) {
    const anchorPos = findLastCharacterPosition(initialAnchorSpinButton);
    if (anchorPos) {
      rootElement.addEventListener(
        "mousemove",
        () => {
          const sel = globalThis.getSelection();
          if (!sel || !sel.anchorNode) {
            return;
          }
          if (sel.anchorNode.nodeType === Node.TEXT_NODE) {
            return;
          }
          const focusNode = sel.focusNode;
          const focusOffset = sel.focusOffset ?? 0;
          if (!focusNode) {
            return;
          }
          const anchorBeforeFocus = compareSelectionPoints(
            anchorPos.node,
            anchorPos.offset,
            focusNode,
            focusOffset
          ) <= 0;
          try {
            sel.removeAllRanges();
            if (anchorBeforeFocus) {
              sel.setBaseAndExtent(
                anchorPos.node,
                anchorPos.offset,
                focusNode,
                focusOffset
              );
            } else {
              sel.setBaseAndExtent(
                focusNode,
                focusOffset,
                anchorPos.node,
                anchorPos.offset
              );
            }
          } catch (error) {
            warnSelectionFailed(
              "selection anchor correction during drag failed",
              error
            );
          }
        },
        { signal: signal.signal, capture: true }
      );
    }
  }
  rootElement.addEventListener(
    "mousemove",
    () => {
      const sel = globalThis.getSelection();
      if (!sel || sel.rangeCount === 0) {
        return;
      }
      const range = sel.getRangeAt(0);
      markDragSelectedElements(rootElement, range);
    },
    { signal: signal.signal, capture: true }
  );
  removeEventTarget.addEventListener(
    eventType,
    (event) => {
      const selectionBeforeRestore = globalThis.getSelection();
      const snapshotAnchorNode = selectionBeforeRestore?.anchorNode ?? null;
      const snapshotAnchorOffset = selectionBeforeRestore?.anchorOffset ?? 0;
      const snapshotFocusNode = selectionBeforeRestore?.focusNode ?? null;
      const snapshotFocusOffset = selectionBeforeRestore?.focusOffset ?? 0;
      const hasMultiElementSelection = selectionBeforeRestore !== null && selectionBeforeRestore.toString().length > 0 && (snapshotAnchorNode !== snapshotFocusNode || snapshotAnchorNode !== null && snapshotAnchorNode === snapshotFocusNode && snapshotAnchorOffset !== snapshotFocusOffset && snapshotAnchorNode.nodeType === Node.ELEMENT_NODE);
      if (hasMultiElementSelection && snapshotAnchorNode && snapshotFocusNode) {
        const anchorIsElementNode = snapshotAnchorNode.nodeType !== Node.TEXT_NODE;
        const anchorNodeWasInContent = anchorIsElementNode ? initialAnchorSpinButton !== null : isEndpointInsideSpinButtonContent(
          snapshotAnchorNode,
          snapshotAnchorOffset
        );
        let anchorSpinButton;
        if (anchorIsElementNode) {
          anchorSpinButton = initialAnchorSpinButton ?? null;
        } else if (anchorNodeWasInContent) {
          anchorSpinButton = resolveSpinButton(
            snapshotAnchorNode,
            rootElement,
            snapshotAnchorOffset
          );
        } else {
          anchorSpinButton = null;
        }
        let anchorNodeForComparison = snapshotAnchorNode;
        let anchorOffsetForComparison = snapshotAnchorOffset;
        if (anchorIsElementNode && anchorSpinButton) {
          const lastPos = findLastCharacterPosition(anchorSpinButton);
          if (lastPos) {
            anchorNodeForComparison = lastPos.node;
            anchorOffsetForComparison = lastPos.offset;
          }
        }
        const focusNodeWasInContent = isEndpointInsideSpinButtonContent(
          snapshotFocusNode,
          snapshotFocusOffset
        );
        const focusSpinButton = focusNodeWasInContent || snapshotFocusNode.nodeType !== Node.TEXT_NODE ? resolveSpinButton(
          snapshotFocusNode,
          rootElement,
          snapshotFocusOffset
        ) : null;
        const anchorBeforeFocus = compareSelectionPoints(
          anchorNodeForComparison,
          anchorOffsetForComparison,
          snapshotFocusNode,
          snapshotFocusOffset
        ) <= 0;
        let anchorPos;
        if (anchorSpinButton?.getAttribute("role") === "spinbutton") {
          anchorPos = anchorBeforeFocus ? findFirstCharacterPosition(anchorSpinButton) : findLastCharacterPosition(anchorSpinButton);
        } else {
          anchorPos = {
            node: anchorNodeForComparison,
            offset: anchorOffsetForComparison
          };
        }
        let focusPos;
        if (focusSpinButton?.getAttribute("role") === "spinbutton") {
          focusPos = anchorBeforeFocus ? findLastCharacterPosition(focusSpinButton) : findFirstCharacterPosition(focusSpinButton);
        } else {
          focusPos = { node: snapshotFocusNode, offset: snapshotFocusOffset };
        }
        if (anchorPos && focusPos) {
          try {
            const range = document.createRange();
            if (anchorBeforeFocus) {
              range.setStart(anchorPos.node, anchorPos.offset);
              range.setEnd(focusPos.node, focusPos.offset);
            } else {
              range.setStart(focusPos.node, focusPos.offset);
              range.setEnd(anchorPos.node, anchorPos.offset);
            }
            const sel = globalThis.getSelection();
            sel?.setBaseAndExtent(
              anchorPos.node,
              anchorPos.offset,
              focusPos.node,
              focusPos.offset
            );
            markDragSelectedElements(rootElement, range);
          } catch (error) {
            warnSelectionFailed(
              "setBaseAndExtent failed when restoring drag selection",
              error
            );
          }
        }
        customRemovalFn?.(event, hasMultiElementSelection);
        rootElement.removeAttribute("data-restoring-selection");
        deferContentEditableRestore(rootElement);
      } else {
        restoreContentEditableAttributes(rootElement);
        if (snapshotAnchorNode && snapshotFocusNode) {
          try {
            globalThis.getSelection()?.setBaseAndExtent(
              snapshotAnchorNode,
              snapshotAnchorOffset,
              snapshotFocusNode,
              snapshotFocusOffset
            );
          } catch (error) {
            warnSelectionFailed(
              "setBaseAndExtent failed when restoring single-element selection",
              error
            );
          }
        }
        customRemovalFn?.(event, hasMultiElementSelection);
        rootElement.removeAttribute("data-restoring-selection");
      }
      signal.abort();
    },
    { signal: signal.signal }
  );
}
function findSpinButtonSnappedPosition(root, anchorNode, anchorOffset, focusNode, focusOffset, direction) {
  let spinButton = null;
  if (focusNode.nodeType === Node.TEXT_NODE) {
    spinButton = focusNode.parentElement?.closest(
      '[role="spinbutton"]'
    ) ?? null;
  } else if (focusNode instanceof HTMLElement && focusNode.getAttribute("role") === "spinbutton") {
    spinButton = focusNode;
  }
  if (!spinButton) {
    return null;
  }
  const allSBs = Array.from(
    root.querySelectorAll('[role="spinbutton"]')
  );
  const idx = allSBs.indexOf(spinButton);
  const focusIsBeforeAnchor = compareSelectionPoints(focusNode, focusOffset, anchorNode, anchorOffset) < 0;
  const isGrowing = direction === "forward" ? !focusIsBeforeAnchor : focusIsBeforeAnchor;
  if (direction === "forward") {
    const nextSB = allSBs[idx + 1] ?? null;
    if (!nextSB) {
      return null;
    }
    if (isGrowing) {
      return findLastCharacterPosition(nextSB) ?? { node: nextSB, offset: 0 };
    }
    return findFirstCharacterPosition(nextSB) ?? { node: nextSB, offset: 0 };
  }
  const prevSB = allSBs[idx - 1] ?? null;
  if (!prevSB) {
    return null;
  }
  if (isGrowing) {
    return findFirstCharacterPosition(prevSB) ?? { node: prevSB, offset: 0 };
  }
  return findLastCharacterPosition(prevSB) ?? { node: prevSB, offset: 0 };
}
function extendSelection(key) {
  const selection = globalThis.getSelection();
  if (selection === null) {
    return;
  }
  const activeElement = globalThis.document?.activeElement;
  const hasMissingEndpoints = selection.anchorNode === null || selection.focusNode === null;
  if (hasMissingEndpoints && isEmptyPlaceholderSpinButton(activeElement)) {
    selection.setBaseAndExtent(
      activeElement,
      0,
      activeElement,
      0
    );
  }
  if (selection.anchorNode === null || selection.focusNode === null || selection.rangeCount < 1) {
    return;
  }
  const anchorNode = selection.anchorNode;
  const anchorOffset = selection.anchorOffset;
  const focusNode = selection.focusNode;
  const focusOffset = selection.focusOffset;
  let nextPos = null;
  const root = focusNode.nodeType === Node.TEXT_NODE ? focusNode.parentElement?.closest(
    ".strato-inputgroup"
  ) : focusNode instanceof HTMLElement ? focusNode.closest(".strato-inputgroup") : null;
  if (root === null || root === void 0) {
    return;
  }
  if (key === "ArrowLeft") {
    nextPos = findSpinButtonSnappedPosition(
      root,
      anchorNode,
      anchorOffset,
      focusNode,
      focusOffset,
      "backward"
    ) ?? findPreviousCharacterPosition(root, focusNode, focusOffset);
  } else if (key === "ArrowRight") {
    nextPos = findSpinButtonSnappedPosition(
      root,
      anchorNode,
      anchorOffset,
      focusNode,
      focusOffset,
      "forward"
    ) ?? findNextCharacterPosition(root, focusNode, focusOffset);
  } else if (key === "Home") {
    nextPos = findFirstCharacterPosition(root);
  } else if (key === "End") {
    nextPos = findLastCharacterPosition(root);
  }
  let crossedSeparator = false;
  while (nextPos && getClosestContentEditable(nextPos.node) === null) {
    crossedSeparator = true;
    const skipped = key === "ArrowLeft" || key === "Home" ? findPreviousCharacterPosition(root, nextPos.node, nextPos.offset) : findNextCharacterPosition(root, nextPos.node, nextPos.offset);
    if (!skipped) {
      break;
    }
    nextPos = skipped;
  }
  if (crossedSeparator && nextPos?.node.nodeType === Node.TEXT_NODE) {
    if (key === "ArrowRight") {
      nextPos = { node: nextPos.node, offset: 0 };
    } else if (key === "ArrowLeft") {
      nextPos = {
        node: nextPos.node,
        offset: nextPos.node.nodeValue?.length ?? 0
      };
    }
  }
  if (!nextPos) {
    return;
  }
  const spinButtonFromNode = (node) => {
    const el = node.nodeType === Node.TEXT_NODE ? node.parentElement : node instanceof HTMLElement ? node : null;
    const sb = el?.closest(SPIN_BUTTON_SELECTOR) ?? null;
    return sb?.getAttribute("role") === "spinbutton" ? sb : null;
  };
  const anchorSB = spinButtonFromNode(anchorNode);
  const anchorCE = getClosestContentEditable(anchorNode);
  const nextCE = getClosestContentEditable(nextPos.node);
  const crossesBoundary = anchorCE !== null && anchorCE !== nextCE;
  if (crossesBoundary) {
    let resolvedAnchorNode = anchorNode;
    let resolvedAnchorOffset = anchorOffset;
    const anchorSpinButtonHasContent = anchorSB !== null && anchorSB.childNodes.length > 0;
    if (anchorNode.nodeType !== Node.TEXT_NODE && anchorSpinButtonHasContent) {
      const anchorBeforeFocus = compareSelectionPoints(
        anchorNode,
        anchorOffset,
        nextPos.node,
        nextPos.offset
      ) <= 0;
      const resolved = anchorBeforeFocus ? findFirstCharacterPosition(anchorSB) : findLastCharacterPosition(anchorSB);
      if (resolved) {
        resolvedAnchorNode = resolved.node;
        resolvedAnchorOffset = resolved.offset;
      }
    }
    root.setAttribute("data-restoring-selection", "");
    for (const editable of root.querySelectorAll(
      "[contenteditable]"
    )) {
      const prev = editable.getAttribute("contenteditable") ?? "";
      editable.removeAttribute("contenteditable");
      editable.dataset.wascontenteditable = prev;
    }
    try {
      selection.setBaseAndExtent(
        resolvedAnchorNode,
        resolvedAnchorOffset,
        nextPos.node,
        nextPos.offset
      );
      const anchorBeforeFocus = compareSelectionPoints(
        resolvedAnchorNode,
        resolvedAnchorOffset,
        nextPos.node,
        nextPos.offset
      ) <= 0;
      const range = document.createRange();
      if (anchorBeforeFocus) {
        range.setStart(resolvedAnchorNode, resolvedAnchorOffset);
        range.setEnd(nextPos.node, nextPos.offset);
      } else {
        range.setStart(nextPos.node, nextPos.offset);
        range.setEnd(resolvedAnchorNode, resolvedAnchorOffset);
      }
      markDragSelectedElements(root, range);
    } catch (error) {
      warnSelectionFailed(
        "setBaseAndExtent failed when extending cross-boundary selection",
        error
      );
    }
    root.removeAttribute("data-restoring-selection");
    deferContentEditableRestore(root);
  } else {
    root.setAttribute("data-restoring-selection", "");
    try {
      selection.setBaseAndExtent(
        anchorNode,
        anchorOffset,
        nextPos.node,
        nextPos.offset
      );
    } catch (error) {
      warnSelectionFailed(
        "setBaseAndExtent failed when extending selection",
        error
      );
    }
    root.removeAttribute("data-restoring-selection");
    const updatedRange = new Range();
    const cmp = anchorNode.compareDocumentPosition(nextPos.node);
    if (cmp & Node.DOCUMENT_POSITION_FOLLOWING || cmp === 0 && anchorOffset <= nextPos.offset) {
      updatedRange.setStart(anchorNode, anchorOffset);
      updatedRange.setEnd(nextPos.node, nextPos.offset);
    } else {
      updatedRange.setStart(nextPos.node, nextPos.offset);
      updatedRange.setEnd(anchorNode, anchorOffset);
    }
    markDragSelectedElements(root, updatedRange);
  }
  root.setAttribute("data-shift-selection", "");
}
function extendSelectionToAll() {
  const selection = globalThis.getSelection();
  if (selection === null || selection.focusNode === null || selection.rangeCount < 1) {
    return;
  }
  const root = selection.focusNode.parentElement?.closest(
    ".strato-inputgroup"
  );
  if (root === null || root === void 0) {
    return;
  }
  const firstPosition = findFirstCharacterPosition(root);
  const lastPosition = findLastCharacterPosition(root);
  if (!firstPosition || !lastPosition) {
    return;
  }
  root.setAttribute("data-restoring-selection", "");
  for (const editable of root.querySelectorAll(
    "[contenteditable]"
  )) {
    const prev = editable.getAttribute("contenteditable") ?? "";
    editable.removeAttribute("contenteditable");
    editable.dataset.wascontenteditable = prev;
  }
  try {
    const range = document.createRange();
    range.setStart(firstPosition.node, firstPosition.offset);
    range.setEnd(lastPosition.node, lastPosition.offset);
    selection.setBaseAndExtent(
      firstPosition.node,
      firstPosition.offset,
      lastPosition.node,
      lastPosition.offset
    );
    markDragSelectedElements(root, range, { includeTextboxes: true });
  } catch (error) {
    warnSelectionFailed(
      "setBaseAndExtent failed when extending selection to all",
      error
    );
  }
  root.removeAttribute("data-restoring-selection");
  root.focus();
  deferContentEditableRestore(root);
}
function getBoundaryNodes(direction, anchorNode, anchorOffset, focusNode, focusOffset) {
  const anchorBeforeFocus = compareSelectionPoints(anchorNode, anchorOffset, focusNode, focusOffset) <= 0;
  if (direction === "left" && anchorBeforeFocus) {
    return { boundaryNode: anchorNode, boundaryOffset: anchorOffset };
  }
  return { boundaryNode: focusNode, boundaryOffset: focusOffset };
}
function collapseMultiElementSelection(rootElement, direction) {
  const selection = globalThis.getSelection();
  if (selection === null || selection.rangeCount < 1 || selection.isCollapsed) {
    return false;
  }
  const anchorNode = selection.anchorNode;
  const focusNode = selection.focusNode;
  if (anchorNode === null || focusNode === null || anchorNode === focusNode) {
    return false;
  }
  if (!rootElement.contains(anchorNode) && !rootElement.contains(focusNode)) {
    return false;
  }
  const anchorOffset = selection.anchorOffset;
  const focusOffset = selection.focusOffset;
  const { boundaryNode, boundaryOffset } = getBoundaryNodes(
    direction,
    anchorNode,
    anchorOffset,
    focusNode,
    focusOffset
  );
  pendingRestoreSignals.get(rootElement)?.abort();
  pendingRestoreSignals.delete(rootElement);
  restoreContentEditableAttributes(rootElement);
  rootElement.removeAttribute("data-shift-selection");
  const target = resolveSpinButton(boundaryNode, rootElement, boundaryOffset);
  selection.removeAllRanges();
  if (target === null) {
    return false;
  }
  target.focus();
  return true;
}
const SPIN_BUTTON_SELECTOR = "[contenteditable], [data-wascontenteditable]";
function getClosestContentEditable(node) {
  return node.nodeType === Node.TEXT_NODE ? node.parentElement?.closest(
    SPIN_BUTTON_SELECTOR
  ) ?? null : node.closest?.(SPIN_BUTTON_SELECTOR) ?? null;
}
function isExpressionNode(node) {
  return node !== null && getClosestContentEditable(node)?.getAttribute("role") === "textbox";
}
function resolveSpinButton(node, root, offset) {
  const direct = getClosestContentEditable(node);
  if (direct !== null) {
    return direct;
  }
  if (node.nodeType !== Node.TEXT_NODE && offset !== void 0 && offset > 0) {
    const childBefore = node.childNodes[offset - 1];
    if (childBefore) {
      const nearest = getNearestPrecedingSpinButton(childBefore, root);
      if (nearest) {
        return nearest;
      }
    }
  }
  return getNearestPrecedingSpinButton(node, root);
}
function getNearestPrecedingSpinButton(node, root) {
  const spinButtons = Array.from(
    root.querySelectorAll(SPIN_BUTTON_SELECTOR)
  );
  if (spinButtons.length === 0) {
    return null;
  }
  let nearest = null;
  for (const spinButton of spinButtons) {
    const position = spinButton.compareDocumentPosition(node);
    if (position & Node.DOCUMENT_POSITION_FOLLOWING) {
      nearest = spinButton;
    }
  }
  return nearest ?? spinButtons[0] ?? null;
}
function isEndpointInsideSpinButtonContent(node, offset) {
  const closest = getClosestContentEditable(node);
  if (!closest || closest.getAttribute("role") !== "spinbutton") {
    return false;
  }
  const firstChild = closest.childNodes[0] ?? null;
  const leadingTextNode = firstChild?.nodeType === Node.TEXT_NODE ? firstChild : firstChild?.firstChild?.nodeType === Node.TEXT_NODE ? firstChild.firstChild : null;
  const isLeadingContentNode = node === firstChild || node === leadingTextNode || node instanceof Element && leadingTextNode !== null && node.contains(leadingTextNode) || node.nodeType === Node.TEXT_NODE && firstChild instanceof Element && firstChild.contains(node);
  const nodeLength = node.nodeType === Node.TEXT_NODE ? node.nodeValue?.length ?? 0 : node.textContent?.length ?? 0;
  const isBoundaryOffset = offset === 0 || offset === nodeLength;
  return !(isLeadingContentNode && isBoundaryOffset);
}
function buildSpinButtonSnapshot(anchorNode, anchorOffset, focusNode, focusOffset, inputRefs) {
  if (!anchorNode || !focusNode) {
    return [];
  }
  const anchorBeforeFocus = compareSelectionPoints(anchorNode, anchorOffset, focusNode, focusOffset) <= 0;
  const selectionStartNode = anchorBeforeFocus ? anchorNode : focusNode;
  const selectionStartOffset = anchorBeforeFocus ? anchorOffset : focusOffset;
  const selectionEndNode = anchorBeforeFocus ? focusNode : anchorNode;
  const selectionEndOffset = anchorBeforeFocus ? focusOffset : anchorOffset;
  const entries = [];
  for (const inputRef of inputRefs) {
    if (!inputRef.current) {
      continue;
    }
    for (const [
      key,
      handler
    ] of inputRef.current.spinButtonHandlers.entries()) {
      const el = handler?.element.current;
      if (el) {
        entries.push({ inputRef, key, element: el });
      }
    }
  }
  return entries.filter(
    ({ element }) => selectionOverlapsSpinButton(
      selectionStartNode,
      selectionStartOffset,
      selectionEndNode,
      selectionEndOffset,
      element
    )
  ).map(({ inputRef, key, element }) => ({
    inputRef,
    key,
    previousText: element.textContent ?? null
  }));
}
function restoreSpinButtonSnapshot(snapshot) {
  for (const { inputRef, key, previousText } of snapshot) {
    const handler = inputRef.current?.spinButtonHandlers.get(key);
    if (!handler) {
      continue;
    }
    if (handler.type === "number") {
      handler.applyText(previousText ?? "", true);
    } else if (previousText !== null && previousText.length > 0) {
      if (handler.element.current) {
        handler.element.current.textContent = previousText;
      }
      handler.searchValue(previousText.charAt(0), true);
    } else {
      handler.clearValue(true);
    }
  }
}
function findExpressionFallbacks(snapshot) {
  const result = /* @__PURE__ */ new Map();
  for (const inputRef of new Set(snapshot.map((e) => e.inputRef))) {
    if (!inputRef.current) {
      continue;
    }
    const clearedKeys = new Set(
      snapshot.filter((e) => e.inputRef === inputRef).map((e) => e.key)
    );
    const survivors = [...inputRef.current.spinButtonHandlers.entries()].filter(
      ([key]) => !clearedKeys.has(key)
    );
    if (survivors.length === 0) {
      result.set(inputRef, "");
    } else if (survivors.length === 1) {
      const [, handler] = survivors[0];
      const text = handler?.element.current?.textContent ?? "";
      if (text.length > 0) {
        result.set(inputRef, text);
      }
    }
  }
  return result;
}
function getFirstSnapshotHandler(snapshot) {
  const first = snapshot[0];
  if (!first) {
    return void 0;
  }
  const spinButtonHandlers = first.inputRef.current?.spinButtonHandlers;
  if (!spinButtonHandlers) {
    return void 0;
  }
  const clearedInFirstInput = snapshot.filter(
    (e) => e.inputRef === first.inputRef
  ).length;
  const switchedToRelative = clearedInFirstInput >= spinButtonHandlers.size - 1;
  if (switchedToRelative) {
    return void 0;
  }
  return spinButtonHandlers.get(first.key);
}
function getLastSnapshotHandler(snapshot) {
  const last = snapshot[snapshot.length - 1];
  if (!last) {
    return void 0;
  }
  const spinButtonHandlers = last.inputRef.current?.spinButtonHandlers;
  if (!spinButtonHandlers) {
    return void 0;
  }
  const clearedInLastInput = snapshot.filter(
    (e) => e.inputRef === last.inputRef
  ).length;
  const switchedToRelative = clearedInLastInput >= spinButtonHandlers.size - 1;
  if (switchedToRelative) {
    return void 0;
  }
  return spinButtonHandlers.get(last.key);
}
function findFirstExpressionFocusTarget(inputRefs, affectedExpressionRefs, expressionFallbacks) {
  const allAffected = /* @__PURE__ */ new Set([
    ...affectedExpressionRefs,
    ...expressionFallbacks.keys()
  ]);
  return inputRefs.find((ref) => allAffected.has(ref));
}
function getPartialExpressionElement(selection, inputRefs) {
  for (const ref of inputRefs) {
    const expressionElement = ref.current?.expressionInputRef;
    if (expressionElement && expressionElement.contains(selection.anchorNode) && expressionElement.contains(selection.focusNode) && selection.toString().length < (expressionElement.textContent?.length ?? 0)) {
      return expressionElement;
    }
  }
  return null;
}
function getCrossExpressionElements(selection, inputRefs) {
  if (selection.rangeCount < 1) {
    return null;
  }
  const range = selection.getRangeAt(0);
  const startEl = inputRefs.find(
    (ref) => ref.current?.expressionInputRef?.contains(range.startContainer)
  )?.current?.expressionInputRef;
  const endEl = inputRefs.find(
    (ref) => ref.current?.expressionInputRef?.contains(range.endContainer)
  )?.current?.expressionInputRef;
  if (!startEl && !endEl) {
    return null;
  }
  if (startEl && !endEl) {
    return [startEl, startEl];
  }
  if (!startEl && endEl) {
    return [endEl, endEl];
  }
  if (!startEl || !endEl) {
    return null;
  }
  if (startEl === endEl) {
    return null;
  }
  const startFirst = !!(startEl.compareDocumentPosition(endEl) & Node.DOCUMENT_POSITION_FOLLOWING);
  return startFirst ? [startEl, endEl] : [endEl, startEl];
}
function focusExpression(expressionElement, range) {
  const root = expressionElement.closest(".strato-inputgroup");
  root?.setAttribute("data-restoring-selection", "");
  expressionElement.focus();
  const sel = globalThis.getSelection();
  if (sel) {
    if (!range) {
      range = document.createRange();
      range.selectNodeContents(expressionElement);
      range.collapse(false);
    }
    sel.setBaseAndExtent(
      range.startContainer,
      range.startOffset,
      range.endContainer,
      range.endOffset
    );
  }
  root?.removeAttribute("data-restoring-selection");
}
function getOffsetWithinExpression(expressionElement, container, offset) {
  const range = document.createRange();
  range.selectNodeContents(expressionElement);
  range.setEnd(container, offset);
  return range.toString().length;
}
function getExpressionCursorOffset(expressionElement) {
  const sel = window.getSelection();
  if (!sel?.rangeCount) {
    return void 0;
  }
  const range = sel.getRangeAt(0);
  if (!expressionElement.contains(range.endContainer)) {
    return void 0;
  }
  return getOffsetWithinExpression(
    expressionElement,
    range.endContainer,
    range.endOffset
  );
}
function restoreExpressionCursorOffset(expressionElement, offset) {
  const textNode = expressionElement.firstChild;
  const sel = window.getSelection();
  if (textNode instanceof Text && sel) {
    const range = document.createRange();
    range.setStart(textNode, Math.min(offset, textNode.length));
    range.collapse(true);
    sel.removeAllRanges();
    sel.addRange(range);
  }
}
function deletePartialExpression(selection, expressionElement) {
  const range = selection.getRangeAt(0);
  const text = expressionElement.textContent ?? "";
  const start = expressionElement.contains(range.startContainer) ? getOffsetWithinExpression(
    expressionElement,
    range.startContainer,
    range.startOffset
  ) : 0;
  const end = expressionElement.contains(range.endContainer) ? getOffsetWithinExpression(
    expressionElement,
    range.endContainer,
    range.endOffset
  ) : text.length;
  const from = Math.min(start, end);
  const to = Math.max(start, end);
  const nextText = `${text.slice(0, from)}${text.slice(to)}`;
  expressionElement.textContent = nextText;
  const nextRange = document.createRange();
  const node = expressionElement.firstChild;
  if (node?.nodeType === Node.TEXT_NODE) {
    nextRange.setStart(node, Math.min(from, nextText.length));
  } else {
    nextRange.setStart(expressionElement, 0);
  }
  nextRange.collapse(true);
  focusExpression(expressionElement, nextRange);
  expressionElement.dispatchEvent(new InputEvent("input", { bubbles: true }));
}
function planBulkClear(selection, inputRefs, endpointsOverride) {
  const validOverride = endpointsOverride?.anchorNode.isConnected && endpointsOverride?.focusNode.isConnected ? endpointsOverride : void 0;
  const anchorNode = validOverride?.anchorNode ?? selection.anchorNode;
  const anchorOffset = validOverride?.anchorOffset ?? selection.anchorOffset;
  const focusNode = validOverride?.focusNode ?? selection.focusNode;
  const focusOffset = validOverride?.focusOffset ?? selection.focusOffset;
  const affectedExpressionRefs = inputRefs.filter((ref) => {
    const expressionInputRef = ref.current?.expressionInputRef;
    return expressionInputRef && (expressionInputRef.contains(anchorNode) || expressionInputRef.contains(focusNode));
  });
  const anchorInExpression = affectedExpressionRefs.some(
    (ref) => ref.current?.expressionInputRef?.contains(anchorNode)
  );
  const focusInExpression = affectedExpressionRefs.some(
    (ref) => ref.current?.expressionInputRef?.contains(focusNode)
  );
  const snapshot = anchorInExpression && focusInExpression ? [] : buildSpinButtonSnapshot(
    anchorNode,
    anchorOffset,
    focusNode,
    focusOffset,
    inputRefs
  );
  return {
    snapshot,
    affectedExpressionRefs,
    expressionFallbacks: findExpressionFallbacks(snapshot)
  };
}
function deleteCrossExpressionPartial(selection, [firstExpressionElement, lastExpressionElement]) {
  if (firstExpressionElement === lastExpressionElement) {
    deletePartialExpression(selection, firstExpressionElement);
    return;
  }
  const { startContainer, startOffset, endContainer, endOffset } = selection.getRangeAt(0);
  const fromRange = document.createRange();
  fromRange.setStart(startContainer, startOffset);
  fromRange.setEnd(
    firstExpressionElement,
    firstExpressionElement.childNodes.length
  );
  fromRange.deleteContents();
  const toRange = document.createRange();
  toRange.setStart(lastExpressionElement, 0);
  toRange.setEnd(endContainer, endOffset);
  toRange.deleteContents();
  selection.removeAllRanges();
  focusExpression(firstExpressionElement);
  firstExpressionElement.dispatchEvent(
    new InputEvent("input", { bubbles: true })
  );
  lastExpressionElement.dispatchEvent(
    new InputEvent("input", { bubbles: true })
  );
}
function restoreFocusAfterBulkClear(firstSpinButtonHandler, focusTarget) {
  if (firstSpinButtonHandler) {
    firstSpinButtonHandler.element.current?.focus();
    firstSpinButtonHandler.selectAll();
    return;
  }
  const expressionElement = focusTarget?.current?.expressionInputRef;
  if (expressionElement) {
    requestAnimationFrame(() => {
      focusExpression(expressionElement);
    });
  }
}
