// @dynatrace/strato-components/forms/date-time-range-picker/selection-copy-helpers.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var selection_copy_helpers_exports = {};
__export(selection_copy_helpers_exports, {
  buildTimeframeClipboardText: () => buildTimeframeClipboardText,
  convertCopyString: () => convertCopyString,
  getPatternAndSelectedText: () => getPatternAndSelectedText,
  onTimeframeCopyCapture: () => onTimeframeCopyCapture
});
module.exports = __toCommonJS(selection_copy_helpers_exports);
var import_tz = require("@date-fns/tz");
var import_date_fns = require("date-fns");
var import_units = require("@dynatrace-sdk/units");
var import_user_preferences = require("@dynatrace-sdk/user-preferences");
var import_selection_helpers = require("./selection-helpers.js");
var import_timeParser = require("../forms-core/time/timeParser.js");
function getPlaceholder(el) {
  const raw = el.style.getPropertyValue("--data-placeholder").trim();
  return raw.startsWith('"') && raw.endsWith('"') ? raw.slice(1, -1) : raw;
}
function sanitizeCopyString(selectionString) {
  return selectionString.split("\u2192").map((part) => part.replaceAll("\n", "").trim()).join(" ").trim();
}
function convertCopyString(selectionString, patterns) {
  const matchableDateTime = selectionString.split("\u2192").map((dateTime) => dateTime.replace(/\n/g, "").trim());
  const parsedDateTimes = [];
  let hasNonISO = false;
  for (const [index, dateTimeInput] of matchableDateTime.entries()) {
    if ((0, import_timeParser.isExpressionString)(dateTimeInput)) {
      parsedDateTimes[index] = (0, import_units.parseTimeAsTimeValue)(dateTimeInput)?.value ?? dateTimeInput;
      continue;
    }
    const currentPattern = patterns?.[index];
    if (currentPattern === void 0 || (currentPattern.includes("dd") && currentPattern.includes("MM") && currentPattern.includes("yyyy")) === false) {
      parsedDateTimes[index] = dateTimeInput;
      hasNonISO = true;
      continue;
    }
    try {
      const dateTimeFormat = (0, import_date_fns.parse)(
        dateTimeInput,
        currentPattern,
        new import_tz.TZDateMini(0, (0, import_user_preferences.getTimezone)())
      );
      let precisionPattern = "yyyy-MM-dd'T'HH:mm:ss.SSSXXX";
      if (currentPattern.includes("ss")) {
        precisionPattern = "yyyy-MM-dd'T'HH:mm:ssXXX";
      } else if (currentPattern.includes("mm")) {
        precisionPattern = "yyyy-MM-dd'T'HH:mmXXX";
      } else if (currentPattern.includes("dd") && currentPattern.includes("MM") && currentPattern.includes("yyyy")) {
        precisionPattern = "yyyy-MM-dd";
      }
      parsedDateTimes[index] = (0, import_date_fns.format)(dateTimeFormat, precisionPattern);
    } catch {
      parsedDateTimes[index] = dateTimeInput;
      hasNonISO = true;
    }
  }
  if (parsedDateTimes.length > 0) {
    const separator = matchableDateTime.length > 1 && hasNonISO ? "\u2192" : "/";
    return parsedDateTimes.filter(Boolean).join(separator);
  }
  return sanitizeCopyString(selectionString);
}
function getPatternAndSelectedText(selection) {
  if (selection.rangeCount === 0) {
    return { pattern: null, text: "" };
  }
  const range = selection.getRangeAt(0);
  const rawStartNode = range.startContainer;
  const rawEndNode = range.endContainer;
  const root = rawStartNode.parentElement?.closest(".strato-inputgroup");
  if (root === null || root === void 0) {
    return { pattern: null, text: "" };
  }
  const patternParts = [];
  let text = "";
  let capturing = false;
  const treeWalker = document.createTreeWalker(
    root,
    NodeFilter.SHOW_TEXT | NodeFilter.SHOW_ELEMENT,
    {
      acceptNode: (node) => {
        if (node.nodeType === Node.TEXT_NODE) {
          return NodeFilter.FILTER_ACCEPT;
        }
        if (node instanceof Element && node.getAttribute("role") === "spinbutton" && node.childNodes.length === 0) {
          return NodeFilter.FILTER_ACCEPT;
        }
        return NodeFilter.FILTER_SKIP;
      }
    }
  );
  let currentNode = treeWalker.nextNode();
  const startNode = rawStartNode.nodeType === Node.TEXT_NODE ? rawStartNode : rawStartNode?.firstChild ?? rawStartNode ?? null;
  const endNode = rawEndNode.nodeType === Node.TEXT_NODE ? rawEndNode : rawEndNode?.firstChild ?? rawEndNode ?? null;
  while (currentNode) {
    if (currentNode === startNode) {
      capturing = true;
    }
    if (capturing && (0, import_selection_helpers.checkIfValueTypeAlignment)(currentNode)) {
      if (currentNode.nodeType === Node.TEXT_NODE) {
        text += currentNode.textContent ?? "";
        const parentElement = currentNode.parentElement;
        if (parentElement?.hasAttribute("data-pattern")) {
          patternParts.push(parentElement.getAttribute("data-pattern") ?? "");
        } else if (parentElement?.dataset.valuetype !== "expression") {
          patternParts.push(currentNode.textContent ?? "");
        }
      } else {
        const el = currentNode;
        const placeholder = getPlaceholder(el);
        text += placeholder;
        if (el.dataset.pattern !== void 0) {
          patternParts.push(el.dataset.pattern);
        } else if (el.dataset.valuetype !== "expression") {
          patternParts.push(placeholder);
        }
      }
    }
    if (currentNode === endNode) {
      break;
    }
    currentNode = treeWalker.nextNode();
  }
  const pattern = patternParts.length > 0 ? patternParts.join("").split("\u2192") : null;
  return { pattern, text };
}
function buildTimeframeClipboardText() {
  const selection = window.getSelection();
  if (selection === null || selection.rangeCount === 0) {
    return null;
  }
  const range = selection.getRangeAt(0);
  const root = range.startContainer.parentElement?.closest(
    ".strato-inputgroup"
  );
  if (root === null || root === void 0) {
    return null;
  }
  const { text: selectionString, pattern } = getPatternAndSelectedText(selection);
  return convertCopyString(selectionString, pattern);
}
function onTimeframeCopyCapture(clipboardEvent) {
  const clipBoardText = buildTimeframeClipboardText();
  if (clipBoardText === null) {
    return;
  }
  clipboardEvent.preventDefault();
  clipboardEvent.clipboardData?.setData("text/plain", clipBoardText);
}
