// @dynatrace/strato-components/forms/date-time-range-picker/useSelectionInteractionHandlers.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useSelectionInteractionHandlers_exports = {};
__export(useSelectionInteractionHandlers_exports, {
  useSelectionInteractionHandlers: () => useSelectionInteractionHandlers
});
module.exports = __toCommonJS(useSelectionInteractionHandlers_exports);
var import_react = require("react");
var import_expression_trigger_helpers = require("./expression-trigger-helpers.js");
var import_selection_copy_helpers = require("./selection-copy-helpers.js");
var import_selection_helpers = require("./selection-helpers.js");
function getInputRefs(dateInputFromRef, dateInputToRef) {
  return [dateInputFromRef, ...dateInputToRef ? [dateInputToRef] : []];
}
function useSelectionInteractionHandlers(touchedContainerRef, onFocusInput, dateInputFromRef, dateInputToRef) {
  const onMouseDown = (0, import_react.useCallback)(
    (event) => {
      if (event.button !== 0) {
        return;
      }
      const root = event.currentTarget;
      const triggerElement = event.target;
      const clickedOnSpinButton = triggerElement instanceof HTMLElement && triggerElement.hasAttribute("contenteditable");
      let initialAnchorSpinButton = null;
      if (!clickedOnSpinButton) {
        const spinButtons = Array.from(
          root.querySelectorAll("[contenteditable]")
        );
        let bestRight = -Infinity;
        for (const sb of spinButtons) {
          const rect = sb.getBoundingClientRect();
          if (rect.right <= event.clientX && rect.right > bestRight) {
            initialAnchorSpinButton = sb;
            bestRight = rect.right;
          }
        }
        if (initialAnchorSpinButton === null && spinButtons.length > 0) {
          initialAnchorSpinButton = spinButtons[0] ?? null;
        }
      }
      window.getSelection()?.removeAllRanges();
      if (root) {
        (0, import_selection_helpers.removeContentEditable)(
          root,
          document,
          "mouseup",
          (mouseUpEvent, hasMultiElementSelection) => {
            if (!hasMultiElementSelection && mouseUpEvent.target instanceof HTMLElement && mouseUpEvent.target === triggerElement) {
              if (mouseUpEvent.target.hasAttribute("contenteditable")) {
                mouseUpEvent.target.focus();
              } else {
                onFocusInput();
              }
            }
          },
          initialAnchorSpinButton
        );
      }
    },
    [onFocusInput]
  );
  const onKeyDown = (0, import_react.useCallback)(
    (event) => {
      const hasModifierKey = event.metaKey === true || event.ctrlKey === true;
      const pickerRoot = event.currentTarget;
      if (event.key === "Shift" || event.key === "Control" || event.key === "Meta" || event.key === "Alt" || event.key === "Command") {
        event.preventDefault();
        return;
      }
      function applyBulkClear(plan, selection) {
        const { snapshot, affectedExpressionRefs, expressionFallbacks } = plan;
        for (const { inputRef, key } of snapshot) {
          inputRef.current?.spinButtonHandlers.get(key)?.clearValue(true);
        }
        for (const [inputRef, text] of expressionFallbacks) {
          inputRef.current?.setExpressionValue(text);
        }
        for (const ref of affectedExpressionRefs) {
          ref.current?.setExpressionValue("");
        }
        selection.removeAllRanges();
      }
      function bulkClear(selection, direction2) {
        const inputRefs = getInputRefs(dateInputFromRef, dateInputToRef);
        const partialExpressionElement = (0, import_selection_helpers.getPartialExpressionElement)(
          selection,
          inputRefs
        );
        if (partialExpressionElement) {
          onBeforeExpressionDelete([partialExpressionElement], inputRefs);
          (0, import_selection_helpers.deletePartialExpression)(selection, partialExpressionElement);
          return;
        }
        const crossExpressionElements = (0, import_selection_helpers.getCrossExpressionElements)(
          selection,
          inputRefs
        );
        if (crossExpressionElements) {
          const isSameExpressionSelection = crossExpressionElements[0] === crossExpressionElements[1];
          const expressionElementsForUndo = isSameExpressionSelection ? [crossExpressionElements[0]] : crossExpressionElements;
          onBeforeExpressionDelete(expressionElementsForUndo, inputRefs);
          const boundaryExpressionRef = isSameExpressionSelection ? inputRefs.find(
            (inputRef) => inputRef.current?.expressionInputRef === crossExpressionElements[0]
          ) : void 0;
          const plan2 = (0, import_selection_helpers.planBulkClear)(
            selection,
            inputRefs,
            (0, import_selection_helpers.consumePreRestoreEndpoints)(pickerRoot)
          );
          const firstSpinButtonHandler = (0, import_selection_helpers.getFirstSnapshotHandler)(plan2.snapshot);
          const expressionFocusTarget2 = (0, import_selection_helpers.findFirstExpressionFocusTarget)(
            inputRefs,
            plan2.affectedExpressionRefs,
            plan2.expressionFallbacks
          );
          (0, import_selection_helpers.deleteCrossExpressionPartial)(selection, crossExpressionElements);
          for (const { inputRef, key } of plan2.snapshot) {
            inputRef.current?.spinButtonHandlers.get(key)?.clearValue(true);
          }
          for (const [inputRef, text] of plan2.expressionFallbacks) {
            if (inputRef !== boundaryExpressionRef) {
              inputRef.current?.setExpressionValue(text);
            }
          }
          (0, import_selection_helpers.restoreFocusAfterBulkClear)(
            firstSpinButtonHandler,
            expressionFocusTarget2
          );
          return;
        }
        const plan = (0, import_selection_helpers.planBulkClear)(
          selection,
          inputRefs,
          (0, import_selection_helpers.consumePreRestoreEndpoints)(pickerRoot)
        );
        const nextFocusedElementHandler = direction2 === "backward" ? (0, import_selection_helpers.getFirstSnapshotHandler)(plan.snapshot) : (0, import_selection_helpers.getLastSnapshotHandler)(plan.snapshot);
        const expressionFocusTarget = (0, import_selection_helpers.findFirstExpressionFocusTarget)(
          inputRefs,
          plan.affectedExpressionRefs,
          plan.expressionFallbacks
        );
        applyBulkClear(plan, selection);
        (0, import_selection_helpers.restoreFocusAfterBulkClear)(
          nextFocusedElementHandler,
          expressionFocusTarget
        );
      }
      function onBeforeExpressionDelete(expressionElements, inputRefs) {
        const expressionSnapshots = [];
        for (const exprEl of expressionElements) {
          let matchingRef;
          for (const inputRef of inputRefs) {
            if (inputRef.current?.expressionInputRef === exprEl) {
              matchingRef = inputRef;
              break;
            }
          }
          if (matchingRef) {
            expressionSnapshots.push({
              inputRef: matchingRef,
              previousText: matchingRef.current?.inputRef?.value ?? exprEl.textContent ?? ""
            });
          }
        }
      }
      if (hasModifierKey && event.key === "a") {
        event.preventDefault();
        if (event.currentTarget instanceof HTMLElement) {
          (0, import_selection_helpers.extendSelectionToAll)();
        }
        return;
      }
      if (event.shiftKey === true && (event.key === "ArrowLeft" || event.key === "ArrowRight" || event.key === "Home" || event.key === "End")) {
        event.preventDefault();
        if (event.currentTarget instanceof HTMLElement) {
          (0, import_selection_helpers.extendSelection)(event.key);
        }
        return;
      }
      const isCut = hasModifierKey && event.key === "x";
      const isDelete = event.key === "Backspace" || event.key === "Delete";
      const direction = isCut || event.key === "Backspace" ? "backward" : "forward";
      if (isCut || isDelete) {
        const selection = window.getSelection();
        if (selection && !selection.isCollapsed) {
          event.preventDefault();
          const clipboardText = (0, import_selection_copy_helpers.buildTimeframeClipboardText)();
          bulkClear(selection, direction);
          if (isCut) {
            if (clipboardText !== null) {
              navigator.clipboard.writeText(clipboardText).catch(() => {
              });
            }
          }
          return;
        }
      }
      if (!hasModifierKey && !event.shiftKey && event.key.length === 1 && ((0, import_expression_trigger_helpers.isExpressionTriggerKey)(event) || /^\d$/.test(event.key))) {
        const selection = window.getSelection();
        if (selection && !selection.isCollapsed && selection.anchorNode !== selection.focusNode) {
          event.preventDefault();
          const inputRefs = getInputRefs(dateInputFromRef, dateInputToRef);
          const plan = (0, import_selection_helpers.planBulkClear)(
            selection,
            inputRefs,
            (0, import_selection_helpers.consumePreRestoreEndpoints)(pickerRoot)
          );
          const firstEntry = plan.snapshot[0];
          const firstKey = firstEntry ? [
            ...firstEntry.inputRef.current?.spinButtonHandlers.keys() ?? []
          ][0] : void 0;
          const expressionModeWillSwitch = plan.expressionFallbacks.size > 0 && firstEntry?.key === firstKey;
          if (expressionModeWillSwitch) {
            const firstSpinButtonHandler = (0, import_selection_helpers.getFirstSnapshotHandler)(
              plan.snapshot
            );
            const expressionFocusTarget = (0, import_selection_helpers.findFirstExpressionFocusTarget)(
              inputRefs,
              plan.affectedExpressionRefs,
              plan.expressionFallbacks
            );
            applyBulkClear(plan, selection);
            expressionFocusTarget?.current?.setExpressionValue(event.key);
            (0, import_selection_helpers.restoreFocusAfterBulkClear)(
              firstSpinButtonHandler,
              expressionFocusTarget
            );
            return;
          }
          for (const { inputRef, key } of plan.snapshot) {
            inputRef.current?.spinButtonHandlers.get(key)?.clearValue(true);
          }
          selection.removeAllRanges();
          const handler = firstEntry ? firstEntry.inputRef.current?.spinButtonHandlers.get(
            firstEntry.key
          ) : void 0;
          if (handler) {
            handler.element.current?.focus();
            if (handler.type === "number") {
              handler.applyText(event.key, true);
              const newLength = handler.element.current?.textContent?.length ?? 0;
              handler.setSelection({ start: newLength, end: newLength });
            } else {
              handler.element.current?.dispatchEvent(
                new KeyboardEvent("keydown", {
                  key: event.key,
                  bubbles: true,
                  cancelable: true
                })
              );
            }
          }
          return;
        }
      }
      if (!event.shiftKey && !hasModifierKey && (event.key === "ArrowUp" || event.key === "ArrowDown") && event.target === touchedContainerRef.current) {
        const anchorSpinButton = getAnchorSpinButton();
        if (anchorSpinButton !== null) {
          event.preventDefault();
          anchorSpinButton.focus();
          anchorSpinButton.dispatchEvent(
            new KeyboardEvent("keydown", {
              key: event.key,
              bubbles: true,
              cancelable: true
            })
          );
          return;
        }
      }
      if (event.target === touchedContainerRef.current && !hasModifierKey) {
        const anchorSpinButton = getAnchorSpinButton();
        if (anchorSpinButton !== null) {
          anchorSpinButton.focus();
        } else {
          onFocusInput();
        }
      }
    },
    [onFocusInput, touchedContainerRef, dateInputFromRef, dateInputToRef]
  );
  return { onMouseDown, onKeyDown };
}
function getAnchorSpinButton() {
  const anchorNode = window.getSelection()?.anchorNode ?? null;
  if (anchorNode === null) {
    return null;
  }
  return anchorNode.nodeType === Node.TEXT_NODE ? anchorNode.parentElement?.closest(
    "[contenteditable]"
  ) ?? null : anchorNode?.closest?.("[contenteditable]") ?? null;
}
