// @dynatrace/strato-components/esm/charts/core/components/sparkline-display/SparklineDisplay.js
import { jsx } from "react/jsx-runtime";
import { isNil } from "lodash-es";
import { SparklineContainer } from "./SparklineContainer.js";
import { getWrapperStyle } from "./utils/get-wrapper-style.js";
import { useBehavioralTrackingProps } from "../../../../core/hooks/useBehavioralTrackingProps.js";
import { defaultChartSize } from "../../../../core/styles/defaultChartValues.sty.js";
import { ChartStateHandler } from "../chart-states/ChartStateHandler.js";
import { getChartStatesTemplates } from "../chart-states/utils/get-chart-states-template.js";
const COMPONENT_NAME = "Sparkline";
const SparklineDisplay = (props) => {
  const {
    data,
    loading = false,
    children,
    width,
    height,
    style,
    className,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "data-testid": dataTestId
  } = props;
  const [behavioralTrackingProps] = useBehavioralTrackingProps(props);
  const series = Array.isArray(data) ? data[0] : data;
  const { emptyState, errorState } = getChartStatesTemplates(children);
  const isEmpty = isNil(series) || series.datapoints?.length === 0;
  const wrapperStyle = getWrapperStyle(width, height, style);
  return /* @__PURE__ */ jsx(
    ChartStateHandler,
    {
      ...behavioralTrackingProps,
      "data-testid": dataTestId ?? "sparkline-state-handler",
      empty: isEmpty,
      height: "100%",
      emptyState,
      errorState,
      isLoading: loading,
      className: className ?? defaultChartSize,
      style: wrapperStyle,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "data-dt-component": COMPONENT_NAME,
      children: /* @__PURE__ */ jsx(SparklineContainer, { ...props, data: series })
    }
  );
};
SparklineDisplay["displayName"] = "SparklineDisplay";
export {
  SparklineDisplay
};
