// @dynatrace/strato-components/esm/charts/core/components/sparkline-display/SparklineRenderer.js
import { jsx } from "react/jsx-runtime";
import {
  barHeightAccessor,
  barWidthAccessor,
  barXAxisAccessor,
  barYAxisAccessor,
  definedAccessor,
  xAxisAccessor,
  y0Accessor,
  y1Accessor,
  yAccessor
} from "./utils/accessors.js";
import { shouldDisplayScatterPlot } from "./utils/should-display-scatter-plot.js";
import { AreaCanvas } from "../canvas/components/AreaCanvas.js";
import { BarCanvas } from "../canvas/components/BarCanvas.js";
import { LineCanvas } from "../canvas/components/LineCanvas.js";
const SparklineRenderer = ({
  data,
  width,
  height,
  xScale,
  yScale,
  fillColor,
  strokeColor,
  variant,
  strokeWidth,
  curve,
  radius,
  gapPolicy
}) => {
  const areaColor = typeof fillColor === "string" ? () => fillColor : () => ({ context }) => fillColor({ context });
  return variant === "area" ? /* @__PURE__ */ jsx(
    AreaCanvas,
    {
      data,
      canvasWidth: width,
      canvasHeight: height,
      x: xAxisAccessor(xScale),
      y1: y1Accessor(yScale),
      y0: y0Accessor(yScale),
      defined: definedAccessor,
      color: areaColor,
      displayScatterPlot: shouldDisplayScatterPlot(gapPolicy),
      lineColor: () => strokeColor,
      curve
    }
  ) : variant === "line" ? /* @__PURE__ */ jsx(
    LineCanvas,
    {
      data,
      canvasWidth: width,
      canvasHeight: height,
      x: xAxisAccessor(xScale),
      y: yAccessor(yScale),
      defined: definedAccessor,
      color: () => strokeColor,
      displayScatterPlot: shouldDisplayScatterPlot(gapPolicy),
      strokeWidth,
      curve,
      radius
    }
  ) : /* @__PURE__ */ jsx(
    BarCanvas,
    {
      data,
      canvasWidth: width,
      canvasHeight: height,
      x: barXAxisAccessor(xScale),
      y: barYAxisAccessor(yScale),
      width: (point) => barWidthAccessor(point, xScale),
      height: (point) => barHeightAccessor(point, yScale),
      strokeWidth: 1,
      color: () => strokeColor,
      opacity: () => 1
    }
  );
};
export {
  SparklineRenderer
};
