// @dynatrace/strato-components/charts/annotations-chart/components/axis/AnnotationsAxis.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var AnnotationsAxis_exports = {};
__export(AnnotationsAxis_exports, {
  AnnotationsAxis: () => AnnotationsAxis
});
module.exports = __toCommonJS(AnnotationsAxis_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_useAnnotationsAxisConfig = require("./hooks/useAnnotationsAxisConfig.js");
var import_annotations_x_axis_props = require("./types/annotations-x-axis-props.js");
var import_TimeBottomAxis = require("../../../core/components/axes/components/TimeBottomAxis.js");
var import_ValueBottomAxis = require("../../../core/components/axes/components/ValueBottomAxis.js");
var import_formatter_by_ticks = require("../../../core/utils/formatter-by-ticks.js");
var import_get_formatter = require("../../../core/utils/get-formatter.js");
var import_ticks = require("../../../core/utils/ticks.js");
var import_constants = require("../../constants.js");
const AnnotationsAxis = ({
  scale,
  width,
  height,
  labelsOffset
}) => {
  const config = (0, import_useAnnotationsAxisConfig.useAnnotationsAxisConfig)();
  const horizontalNumTicks = (0, import_ticks.getHorizontalNumTicks)(width);
  const hidden = config?.hidden ?? import_constants.DEFAULT_AXIS_HIDDEN;
  let formatterResult;
  if (config && (0, import_annotations_x_axis_props.isNumericalAxis)(config)) {
    const formatter = (0, import_get_formatter.getFormatter)(config.formatter);
    formatterResult = (0, import_formatter_by_ticks.getFormattedTicks)(
      horizontalNumTicks,
      scale,
      config.unit,
      formatter
    );
  } else {
    formatterResult = (0, import_formatter_by_ticks.getFormattedTicks)(
      horizontalNumTicks,
      scale,
      void 0,
      void 0
    );
  }
  const commonProps = {
    containerWidth: width - labelsOffset,
    containerHeight: height,
    ticks: formatterResult.ticks,
    formattedTicks: formatterResult.formattedTicks,
    label: config?.label
  };
  return !hidden ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      style: {
        marginLeft: `${labelsOffset}px`
      },
      children: (0, import_annotations_x_axis_props.isNumericalScale)(scale) ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ValueBottomAxis.ValueBottomAxis, { scale, ...commonProps }) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TimeBottomAxis.TimeBottomAxis, { scale, ...commonProps })
    }
  ) : null;
};
