// @dynatrace/strato-components/esm/filters/filter-field/editor-state-configs/editor-config.js
import { closeBrackets, closeBracketsKeymap } from "@codemirror/autocomplete";
import { defaultKeymap, history, redo, undo } from "@codemirror/commands";
import { syntaxHighlighting } from "@codemirror/language";
import {
  Annotation,
  Compartment,
  EditorState,
  StateEffect,
  StateField
} from "@codemirror/state";
import { drawSelection, EditorView, keymap } from "@codemirror/view";
import { filterFieldHighlighter } from "../decorations/highlight.js";
import { filterFieldLanguageExtension } from "../parser/filterfield-language.js";
const placeholderCompartment = new Compartment();
const suggestionsCompartment = new Compartment();
const ariaAttributesCompartment = new Compartment();
const disabledPropCompartment = new Compartment();
const readOnlyPropCompartment = new Compartment();
const stringAnnotation = Annotation.define();
const ANNOTATION_DELETE_STATEMENT = "input.type.delete-statement";
const ANNOTATION_SUGGESTION_APPLY = "suggestion.apply";
const ANNOTATION_SUGGESTION_PIN = "suggestion.pin";
const ANNOTATION_COMPOSE_END = "input.type.compose.end";
const ANNOTATION_DOC_CHANGED_CONTROLLED = "doc-changed.controlled";
const ANNOTATION_END_PASTING_MODE = "pasting-mode.end";
const pastingEffect = StateEffect.define();
const pastingMode = StateField.define({
  create: () => {
    return { pastingInProgress: false };
  },
  update(value, transaction) {
    for (const effect of transaction.effects) {
      if (effect.is(pastingEffect)) {
        value = effect.value;
      }
    }
    return value;
  }
});
const validatorMapEffect = StateEffect.define();
const usedValidatorMap = StateField.define({
  create: () => void 0,
  update(value, transaction) {
    for (const effect of transaction.effects) {
      if (effect.is(validatorMapEffect)) {
        value = effect.value;
      }
    }
    return value;
  }
});
const errorsEffect = StateEffect.define();
const usedErrors = StateField.define({
  create() {
    return { errors: [] };
  },
  update(value, transaction) {
    for (const effect of transaction.effects) {
      if (effect.is(errorsEffect)) {
        value.errors = [];
      }
    }
    return value;
  }
});
const overrideKeys = ["ArrowUp", "ArrowDown", "Enter", "Mod-Enter"];
const staticEditorExtensions = [
  /**
   * General
   */
  // Enable line wrapping so that overflowing text is wrapped into the next line.
  EditorView.lineWrapping,
  // Enable multiple selections (CMD / ctrl + drag).
  EditorState.allowMultipleSelections.of(true),
  // Provides a undo/redo history
  history(),
  // Draw custom cursor.
  drawSelection(),
  keymap.of([
    ...defaultKeymap.filter(({ key }) => !overrideKeys.includes(key ?? "")),
    ...closeBracketsKeymap,
    {
      key: "Mod-z",
      run: undo
    },
    {
      key: "Mod-y",
      run: redo
    },
    {
      key: "Mod-Shift-z",
      run: redo
    }
  ]),
  /**
   * Filter field language + syntax highlighting.
   *
   * Error highlighting is also done with the highlight extensions, since using an actual
   * linter, potential errors only show up after a short (configurable) delay. As soon as
   * you type something again, the errors disappear and then show up again if the input is
   * still invalid. Using the decoration plugins, we can immediately show errors.
   * The actual linter extension is only used to show tooltips with proper error messages.
   */
  // The extension for parsing the filter field language into a syntax tree.
  filterFieldLanguageExtension,
  // Automatically close brackets.
  closeBrackets(),
  // Enable highlighting using the defined highlighter.
  // This adds the spans and classes used to style the different parts of a filter.
  syntaxHighlighting(filterFieldHighlighter)
];
const recentFiltersEffect = StateEffect.define();
const usedRecentFilters = StateField.define({
  create() {
    return [];
  },
  update(value, transaction) {
    for (const effect of transaction.effects) {
      if (effect.is(recentFiltersEffect)) {
        value = effect.value;
      }
    }
    return value;
  }
});
const pinnedFiltersEffect = StateEffect.define();
const usedPinnedFilters = StateField.define({
  create() {
    return [];
  },
  update(value, transaction) {
    for (const effect of transaction.effects) {
      if (effect.is(pinnedFiltersEffect)) {
        value = effect.value;
      }
    }
    return value;
  }
});
export {
  ANNOTATION_COMPOSE_END,
  ANNOTATION_DELETE_STATEMENT,
  ANNOTATION_DOC_CHANGED_CONTROLLED,
  ANNOTATION_END_PASTING_MODE,
  ANNOTATION_SUGGESTION_APPLY,
  ANNOTATION_SUGGESTION_PIN,
  ariaAttributesCompartment,
  disabledPropCompartment,
  errorsEffect,
  overrideKeys,
  pastingEffect,
  pastingMode,
  pinnedFiltersEffect,
  placeholderCompartment,
  readOnlyPropCompartment,
  recentFiltersEffect,
  staticEditorExtensions,
  stringAnnotation,
  suggestionsCompartment,
  usedErrors,
  usedPinnedFilters,
  usedRecentFilters,
  usedValidatorMap,
  validatorMapEffect
};
