// @dynatrace/strato-components/esm/filters/filter-field/editor-state-configs/edit-mode-update-listener.js
import { getStatementStatus } from "../utils/get-statement-status.js";
import { getStatementCursorAtPosition } from "../utils/tree-navigation.js";
function handleEditMode(update, suggestionsContext) {
  if (!update.focusChanged) {
    return;
  }
  const { view, state } = update;
  const cursorPos = state.selection.main.anchor;
  const statementStatus = getStatementStatus(
    getStatementCursorAtPosition(cursorPos, state)
  );
  const currentStatement = suggestionsContext.current?.currentStatement;
  const isAtEnd = cursorPos === state.doc.length;
  const hasTrailingSpace = state.sliceDoc(state.doc.length - 1) === " ";
  const hasTrailingBracket = state.sliceDoc(state.doc.length - 1) === ")" && currentStatement?.comparisonOperator?.type !== "InclusionOperator";
  const isStatementComplete = statementStatus === "complete";
  const shouldEnterEditMode = (
    // Focus is moved to the filter field.
    view.hasFocus && !hasTrailingSpace && !hasTrailingBracket && isStatementComplete && isAtEnd
  );
  if (!shouldEnterEditMode) {
    return;
  }
  const transaction = {
    changes: {
      from: state.doc.length,
      to: state.doc.length,
      insert: " "
    },
    selection: { anchor: state.doc.length + 1 }
  };
  view.dispatch(transaction);
}
export {
  handleEditMode
};
