// @dynatrace/strato-components/esm/filters/filter-field/editor-state-configs/pasting-update-listener.js
import {
  ANNOTATION_DOC_CHANGED_CONTROLLED,
  pastingEffect,
  stringAnnotation
} from "./editor-config.js";
function handlePastingTransaction(update, pastingInfoRef, editorView, editorViewRef) {
  const isPastingStart = update.transactions.some(
    (transaction) => transaction.isUserEvent("input.paste")
  );
  const isControlledUpdate = update.transactions.some((transaction) => {
    const annotation = transaction.annotation(stringAnnotation);
    if (annotation === ANNOTATION_DOC_CHANGED_CONTROLLED) {
      return true;
    }
    return false;
  });
  if (isPastingStart) {
    pastingInfoRef.current.pastingInProgress = true;
  }
  if (pastingInfoRef.current.pastingInProgress !== true) {
    return;
  }
  const isPastingEnd = update.transactions.some((transaction) => {
    return transaction.isUserEvent("input.type") || transaction.isUserEvent("delete") || transaction.isUserEvent("select");
  });
  if (isPastingEnd || update.focusChanged || pastingInfoRef.current.cancelPastingMode) {
    pastingInfoRef.current = { pastingInProgress: false };
    editorView.dispatch({
      effects: [
        pastingEffect.of({
          pastingInProgress: false
        })
      ],
      // In order to re-run the decorations and re-generate errors and diagnostics,
      // we have to dispatch a new document.
      changes: {
        from: 0,
        to: editorViewRef.current?.state.doc.length,
        insert: editorViewRef.current?.state.doc
      },
      selection: update.state.selection
    });
    if (pastingInfoRef.current.cancelPastingMode) {
      pastingInfoRef.current.cancelPastingMode = false;
    }
  }
  if (update.docChanged && isPastingStart || isControlledUpdate && pastingInfoRef.current.pastingInProgress === true) {
    const changedRange = isControlledUpdate ? pastingInfoRef.current.range : getChangedRange(update);
    if (changedRange) {
      const { state } = update;
      const range = {
        from: changedRange.from,
        to: changedRange.to
      };
      editorView.dispatch({
        effects: [pastingEffect.of({ pastingInProgress: true, range })],
        // In order to re-run the decorations and re-generate errors and diagnostics,
        // we have to dispatch a new document.
        changes: {
          from: 0,
          to: editorViewRef.current?.state.doc.length,
          insert: editorViewRef.current?.state.doc
        },
        selection: state.selection
      });
    }
  }
}
function getChangedRange(update) {
  const changedRanges = update.changedRanges;
  if (changedRanges && changedRanges.length > 0) {
    if (changedRanges.length === 1) {
      return { from: changedRanges[0].fromB, to: changedRanges[0].toB };
    }
    if (changedRanges.length > 1) {
      const cursorPos = update.state.selection.main.head;
      for (const range of changedRanges) {
        if (range.toB === cursorPos) {
          return { from: range.fromB, to: range.toB };
        }
      }
    }
  }
  return null;
}
function handleControlledPasting(transaction, pastingInfoRef) {
  if (transaction.isUserEvent("input.paste")) {
    const range = { from: 0, to: 0 };
    if (transaction.changes) {
      const insertions = transaction.changes.inserted;
      range.to = transaction.state.selection.main.anchor;
      if (insertions && Array.isArray(insertions) && insertions.length > 0) {
        if (insertions.length > 1) {
          range.from = Math.max(range.to - insertions[1].length, 0);
        } else {
          range.from = Math.max(range.to - insertions[0].length, 0);
        }
      } else {
        range.from = range.to;
      }
    }
    pastingInfoRef.current = { pastingInProgress: true, range };
  } else if (pastingInfoRef.current.pastingInProgress && !transaction.isUserEvent("input.paste")) {
    pastingInfoRef.current.cancelPastingMode = true;
  }
}
export {
  getChangedRange,
  handleControlledPasting,
  handlePastingTransaction
};
