// @dynatrace/strato-components/charts/xy-chart/constants/defaults.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var defaults_exports = {};
__export(defaults_exports, {
  ANNOTATIONS_PADDING_BOTTOM: () => ANNOTATIONS_PADDING_BOTTOM,
  DEFAULT_AXIS_CONFIG: () => DEFAULT_AXIS_CONFIG,
  DEFAULT_CHART_HEIGHT: () => DEFAULT_CHART_HEIGHT,
  DEFAULT_COLOR_PALETTE: () => DEFAULT_COLOR_PALETTE,
  DEFAULT_COLOR_PALETTE_RECT: () => DEFAULT_COLOR_PALETTE_RECT,
  DEFAULT_LEGEND_RANGE_FORMATTER: () => DEFAULT_LEGEND_RANGE_FORMATTER,
  DEFAULT_LOADING_STATE: () => DEFAULT_LOADING_STATE,
  DEFAULT_POINTS_DISPLAY: () => DEFAULT_POINTS_DISPLAY,
  DISPLAY_NAME: () => DISPLAY_NAME,
  DISPLAY_UNITS_LIMIT: () => DISPLAY_UNITS_LIMIT,
  RECT_STROKE_WIDTH: () => RECT_STROKE_WIDTH,
  XY_CHART_ARIA_LABEL: () => XY_CHART_ARIA_LABEL,
  XY_CHART_BOTTOM_AXIS_LABEL: () => XY_CHART_BOTTOM_AXIS_LABEL,
  XY_CHART_TOP_AXIS_LABEL: () => XY_CHART_TOP_AXIS_LABEL,
  defaultConfig: () => defaultConfig,
  defaultLegend: () => defaultLegend,
  defaultToolbar: () => defaultToolbar,
  initialState: () => initialState
});
module.exports = __toCommonJS(defaults_exports);
var import_AnnotationsDataInitialState = require("../../core/components/annotations/providers/AnnotationsDataInitialState.js");
var import_constants = require("../../core/components/legend/range/constants.js");
var import_constants2 = require("../../core/constants.js");
const DEFAULT_CHART_HEIGHT = 300;
const DEFAULT_COLOR_PALETTE = "categorical";
const DEFAULT_COLOR_PALETTE_RECT = "blue";
const DEFAULT_LOADING_STATE = false;
const DEFAULT_AXIS_CONFIG = {
  hidden: false,
  min: "auto",
  max: "auto",
  reversed: false
};
const DEFAULT_LEGEND_RANGE_FORMATTER = {
  maximumFractionDigits: 1
};
const initialState = {
  zoom: {
    zoomInEnabled: true,
    zoomOutEnabled: false
  },
  currentDomain: void 0,
  select: void 0,
  interactionMode: "explore",
  thresholds: { hoveredThresholdId: void 0 },
  highlightedSeriesId: void 0,
  filteredSeriesIds: void 0,
  annotations: import_AnnotationsDataInitialState.initialSelectionState
};
const defaultLegend = {
  hidden: false,
  position: "auto",
  ratio: import_constants2.defaultRatio
};
const defaultToolbar = {
  hidden: false,
  placement: "top-right",
  collapsed: false,
  draggable: true
};
const defaultConfig = {
  legend: defaultLegend,
  toolbar: defaultToolbar,
  colorPalette: void 0,
  truncationMode: import_constants.TRUNCATION_MODE.Middle,
  thresholdIndicators: [],
  yAxes: [],
  xAxes: [],
  colorRules: []
};
const XY_CHART_ARIA_LABEL = "XY Chart";
const ANNOTATIONS_PADDING_BOTTOM = 6;
const DISPLAY_UNITS_LIMIT = 3;
const DEFAULT_POINTS_DISPLAY = "auto";
const RECT_STROKE_WIDTH = 1;
const XY_CHART_BOTTOM_AXIS_LABEL = "Bottom axis";
const XY_CHART_TOP_AXIS_LABEL = "Top axis";
const DISPLAY_NAME = "XYChart";
