// @dynatrace/strato-components/esm/charts/gauge/components/GaugeChartTooltip/GaugeOverlayTooltip.js
import { jsx } from "react/jsx-runtime";
import { useIntl } from "react-intl";
import { OverlayTooltip } from "../../../core/components/overlay-tooltip/OverlayTooltip.js";
import { useTooltipSeriesActions } from "../../../core/components/tooltip/shared/hooks/useTooltipSeriesActions.js";
import { useGaugeChartTooltip } from "../../contexts/GaugeChartTooltipContext.js";
import { useGaugeChartFormatterContext } from "../../hooks/useGaugeChartFormatterContext.js";
const GaugeOverlayTooltip = ({
  value,
  color
}) => {
  const intl = useIntl();
  const { formatter, unit } = useGaugeChartFormatterContext();
  const { name: tooltipName, template } = useGaugeChartTooltip();
  const seriesName = tooltipName ?? intl.formatMessage({ defaultMessage: "Value", id: "jc4XDeI4sf7aLs2W" });
  const formattedValue = formatter(value, unit);
  const tooltipItem = {
    id: `${seriesName}-${value}`,
    name: seriesName,
    values: [formattedValue]
  };
  const seriesActions = useTooltipSeriesActions(
    [tooltipItem],
    (item) => item.id
  );
  const actions = tooltipItem ? () => seriesActions(tooltipItem) : void 0;
  if (template) {
    const customItem = {
      value,
      formattedValue,
      color,
      name: tooltipName,
      unit: typeof unit === "string" ? unit : void 0
    };
    return /* @__PURE__ */ jsx(OverlayTooltip, { legacyTemplate: template(customItem) });
  }
  return /* @__PURE__ */ jsx(OverlayTooltip, { children: /* @__PURE__ */ jsx(
    OverlayTooltip.SeriesItem,
    {
      name: seriesName,
      value: formattedValue,
      color,
      actions
    }
  ) });
};
GaugeOverlayTooltip["displayName"] = "GaugeOverlayTooltip";
export {
  GaugeOverlayTooltip
};
