// @dynatrace/strato-components/esm/buttons/intent-button/IntentButtonWithoutResponse.js
import { jsx, jsxs } from "react/jsx-runtime";
import {
  forwardRef,
  useMemo
} from "react";
import { getIntentLink } from "@dynatrace-sdk/navigation";
import * as styles from "./IntentButton.sty.js";
import { useIntent } from "../../core/hooks/intent.js";
import { AppIcon } from "../../layouts/app-header/app-icon/AppIcon.js";
import { Button } from "../button/Button.js";
const IntentButtonWithoutResponse = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const {
    payload,
    options,
    icon,
    children,
    disabled,
    onClick,
    size,
    textAlign,
    "aria-disabled": ariaDisabledProp,
    ...remainingProps
  } = props;
  const intent = useIntent({
    payload,
    options
  });
  const link = useMemo(() => {
    return getIntentLink(
      payload,
      options.recommendedAppId ?? "",
      options.recommendedIntentId ?? ""
    );
  }, [options.recommendedAppId, options.recommendedIntentId, payload]);
  return /* @__PURE__ */ jsxs(
    Button,
    {
      as: "a",
      href: disabled ? void 0 : link,
      ref: forwardedRef,
      onClick: (event) => {
        onClick?.(event);
        if (event.isDefaultPrevented()) {
          return;
        }
        const nativeEvent = event.nativeEvent;
        if (nativeEvent.button !== 0 || nativeEvent.metaKey || nativeEvent.ctrlKey || nativeEvent.shiftKey) {
          return;
        }
        intent.send();
        event.stopPropagation();
        event.preventDefault();
      },
      onKeyDown: (event) => {
        const nativeEvent = event.nativeEvent;
        if (nativeEvent.code !== "Space" && nativeEvent.code !== "Enter") {
          return;
        }
        event.preventDefault();
        if (nativeEvent.metaKey || nativeEvent.ctrlKey) {
          window.open(link);
          return;
        }
        if (nativeEvent.shiftKey) {
          window.open(link, "_blank", "menubar");
          return;
        }
        intent.send();
      },
      disabled,
      "aria-disabled": ariaDisabledProp,
      textAlign,
      size,
      ...remainingProps,
      children: [
        intent.iconURL && /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(
          AppIcon,
          {
            src: intent.iconURL,
            className: disabled || ariaDisabledProp ? styles.disabledAppIconOpacity : void 0,
            alt: ""
          }
        ) }),
        icon && /* @__PURE__ */ jsx(Button.Prefix, { children: icon }),
        children
      ]
    }
  );
});
export {
  IntentButtonWithoutResponse
};
