// @dynatrace/strato-components/esm/buttons/intent-button/IntentButtonWithResponse.js
import { jsx, jsxs } from "react/jsx-runtime";
import { forwardRef } from "react";
import * as styles from "./IntentButton.sty.js";
import { useIntentWithResponse } from "../../core/hooks/intent.js";
import { AppIcon } from "../../layouts/app-header/app-icon/AppIcon.js";
import { Button } from "../button/Button.js";
const IntentButtonWithResponse = /* @__PURE__ */ forwardRef((props, ref) => {
  const {
    onResponse,
    payload,
    options,
    children,
    icon,
    disabled,
    onClick,
    size,
    textAlign,
    "aria-disabled": ariaDisabledProp,
    ...remainingProps
  } = props;
  const intentWithResponse = useIntentWithResponse({
    payload,
    options
  });
  const handleClick = async (event) => {
    onClick?.(event);
    if (event.isDefaultPrevented()) {
      return;
    }
    try {
      const response = await intentWithResponse.send();
      onResponse(response);
    } catch (error) {
      onResponse({
        status: "error",
        error: new Error("Error sending intent with response", {
          cause: error
        })
      });
    }
  };
  return /* @__PURE__ */ jsxs(
    Button,
    {
      ref,
      onClick: handleClick,
      onMouseDown: (event) => {
        if (event.nativeEvent.button === 1) {
          handleClick(event).catch(() => void 0);
        }
      },
      disabled,
      "aria-disabled": ariaDisabledProp,
      textAlign,
      size,
      ...remainingProps,
      children: [
        intentWithResponse.iconURL && /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(
          AppIcon,
          {
            src: intentWithResponse.iconURL,
            className: disabled || ariaDisabledProp ? styles.disabledAppIconOpacity : void 0,
            alt: ""
          }
        ) }),
        icon && /* @__PURE__ */ jsx(Button.Prefix, { children: icon }),
        children
      ]
    }
  );
});
export {
  IntentButtonWithResponse
};
