// @dynatrace/strato-components/esm/buttons/intent-button/IntentButton.js
import { jsx, jsxs } from "react/jsx-runtime";
import { createElement } from "react";
import { forwardRef, isValidElement } from "react";
import { useIntl } from "react-intl";
import { DotMenuIcon } from "@dynatrace/strato-icons";
import { IntentButtonWithoutResponse } from "./IntentButtonWithoutResponse.js";
import { IntentButtonWithResponse } from "./IntentButtonWithResponse.js";
import { OpenWith } from "./OpenWith.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useSlots } from "../../core/hooks/useSlots.js";
import {
  Intent as Item
} from "../../core/slots/Intent/Intent.js";
import { mergeProps } from "../../core/utils/merge-props.js";
import { InputGroup } from "../../layouts/input-group/InputGroup.js";
import { Menu } from "../../navigation/menu/Menu.js";
import { Prefix as IconPrefix } from "../../navigation/menu/Prefix.js";
import { Button } from "../button/Button.js";
const COMPONENT_NAME = "IntentButton";
const _IntentButton = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const {
    payload,
    children,
    onResponse,
    options,
    disabled,
    color,
    variant = "emphasized",
    width,
    onClick,
    size,
    textAlign,
    showMenu = true,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "aria-disabled": ariaDisabledProp,
    className: consumerClassName,
    style: consumerStyle,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
  const [slots, freeform] = useSlots(children, {
    icon: IconPrefix,
    items: { match: Item, multiple: true },
    openWith: OpenWith
  });
  const slottedIcon = isValidElement(slots.icon) ? slots.icon : void 0;
  const intl = useIntl();
  const menuAriaLabel = intl.formatMessage({
    defaultMessage: "Open intent menu",
    id: "OEhnk3JokTy0c/JX",
    description: "Aria label for the IntentButton menu trigger."
  });
  const sharedProps = {
    color,
    variant,
    width,
    size,
    textAlign,
    onClick,
    disabled,
    "aria-disabled": ariaDisabledProp
  };
  return /* @__PURE__ */ jsxs(
    InputGroup,
    {
      ref: forwardedRef,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      className: consumerClassName,
      style: consumerStyle,
      ...mergeProps(restProps, behavioralTrackingProps),
      "data-dt-component": COMPONENT_NAME,
      children: [
        "responseProperties" in options ? /* @__PURE__ */ jsx(
          IntentButtonWithResponse,
          {
            ...sharedProps,
            payload,
            options,
            onResponse: onResponse ?? (() => void 0),
            icon: slottedIcon,
            children: freeform
          }
        ) : /* @__PURE__ */ jsx(
          IntentButtonWithoutResponse,
          {
            ...sharedProps,
            payload,
            options,
            icon: slottedIcon,
            children: freeform
          }
        ),
        showMenu && /* @__PURE__ */ jsxs(Menu, { children: [
          /* @__PURE__ */ jsx(Menu.Trigger, { "aria-label": menuAriaLabel, children: /* @__PURE__ */ jsx(
            Button,
            {
              disabled,
              "aria-disabled": ariaDisabledProp,
              variant,
              color,
              size,
              style: { width: "max-content", minWidth: "fit-content" },
              children: /* @__PURE__ */ jsx(DotMenuIcon, {})
            }
          ) }),
          /* @__PURE__ */ jsxs(
            Menu.Content,
            {
              alignment: "end",
              "data-dtrum-mask": dataDtrumMask,
              "data-dtrum-allow": dataDtrumAllow,
              children: [
                slots.items && /* @__PURE__ */ jsx(Menu.Group, { children: slots.items.map((item, index) => /* @__PURE__ */ createElement(
                  Menu.Intent,
                  {
                    ...item.props,
                    key: `${item.props.options.recommendedAppId}-${index}`
                  }
                )) }),
                /* @__PURE__ */ jsx(Menu.Group, { children: slots.openWith ?? /* @__PURE__ */ jsx(OpenWith, { payload }) })
              ]
            }
          )
        ] })
      ]
    }
  );
});
_IntentButton.displayName = COMPONENT_NAME;
const IntentButton = Object.assign(_IntentButton, {
  Item,
  OpenWith,
  Icon: IconPrefix
});
export {
  IntentButton
};
