// @dynatrace/strato-components/tables/convert-to-columns/convert-to-columns.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var convert_to_columns_exports = {};
__export(convert_to_columns_exports, {
  convertToColumns: () => convertToColumns
});
module.exports = __toCommonJS(convert_to_columns_exports);
var import_units = require("@dynatrace-sdk/units");
function extractColumnTypes(rangedFieldTypes) {
  const columnTypes = /* @__PURE__ */ new Map();
  for (const { mappings } of rangedFieldTypes) {
    for (const key in mappings) {
      const type = mappings[key]?.type ?? "string";
      if (columnTypes.has(key)) {
        columnTypes.get(key)?.add(type);
      } else {
        columnTypes.set(key, /* @__PURE__ */ new Set([type]));
      }
    }
  }
  return columnTypes;
}
const FieldTypeTypeToColumnType = /* @__PURE__ */ new Map([
  ["timestamp", "datetime"],
  ["long", "long"],
  ["double", "number"],
  ["duration", "number"]
]);
const FieldTypeTypeToFormatter = /* @__PURE__ */ new Map([
  [
    "duration",
    {
      input: import_units.units.time.nanosecond,
      cascade: 3
    }
  ]
]);
function isTimeFrameEntry(value) {
  if (value === null || typeof value !== "object") {
    return false;
  }
  const { start, end } = value;
  return (typeof start === "string" || start instanceof Date) && (typeof end === "string" || end instanceof Date);
}
function isTimeGeoPointEntry(value) {
  if (value === null || typeof value !== "object") {
    return false;
  }
  const { latitude, longitude } = value;
  return typeof latitude === "number" && typeof longitude === "number";
}
function getTimeframeColumn(columnKey) {
  return {
    header: columnKey,
    id: columnKey,
    columns: [
      {
        header: "start",
        id: `${columnKey}.start`,
        accessor: (row) => {
          const entry = row?.[columnKey];
          if (isTimeFrameEntry(entry)) {
            const rawDate = entry.start;
            if (rawDate instanceof Date) {
              return rawDate;
            }
            const date = new Date(rawDate);
            return Number.isNaN(date.getTime()) ? "" : date;
          }
          return "";
        },
        columnType: "datetime",
        width: "content"
      },
      {
        header: "end",
        id: `${columnKey}.end`,
        accessor: (row) => {
          const entry = row?.[columnKey];
          if (isTimeFrameEntry(entry)) {
            const rawDate = entry.end;
            if (rawDate instanceof Date) {
              return rawDate;
            }
            const date = new Date(rawDate);
            return Number.isNaN(date.getTime()) ? "" : date;
          }
          return "";
        },
        columnType: "datetime",
        width: "content"
      }
    ]
  };
}
function getGeoPointColumn(columnKey) {
  return {
    header: columnKey,
    id: columnKey,
    columns: [
      {
        header: "longitude",
        id: `${columnKey}.longitude`,
        accessor: (row) => {
          if (isTimeGeoPointEntry(row?.[columnKey])) {
            return row[columnKey].longitude;
          }
          return "";
        },
        columnType: "text",
        width: "content"
      },
      {
        header: "latitude",
        id: `${columnKey}.latitude`,
        accessor: (row) => {
          if (isTimeGeoPointEntry(row?.[columnKey])) {
            return row[columnKey].latitude;
          }
          return "";
        },
        columnType: "text",
        width: "content"
      }
    ]
  };
}
function convertToColumns(rangedFieldTypes) {
  const extractedTypes = extractColumnTypes(rangedFieldTypes);
  const columns = [];
  for (const [columnKey, columnTypes] of extractedTypes.entries()) {
    const columnHasOnlyOneType = columnTypes.size === 1;
    const columnType = columnHasOnlyOneType ? Array.from(columnTypes)[0] : "string";
    if (columnType === "timeframe") {
      columns.push(getTimeframeColumn(columnKey));
    } else if (columnType === "geo_point") {
      columns.push(getGeoPointColumn(columnKey));
    } else {
      columns.push({
        header: columnKey,
        id: columnKey,
        accessor: (row) => row?.[columnKey],
        width: "content",
        columnType: FieldTypeTypeToColumnType.get(columnType) ?? "text",
        formatter: FieldTypeTypeToFormatter.get(columnType) ?? void 0
      });
    }
  }
  return columns;
}
