// @dynatrace/strato-components/forms/forms-core/components/field-set/FieldSet.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var FieldSet_exports = {};
__export(FieldSet_exports, {
  FieldSet: () => FieldSet
});
module.exports = __toCommonJS(FieldSet_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_is = require("react-is");
var styles = __toESM(require("./FieldSet.sty.js"));
var import_useBehavioralTrackingProps = require("../../../../core/hooks/useBehavioralTrackingProps.js");
var import_merge_props = require("../../../../core/utils/merge-props.js");
var import_parse_boolean = require("../../../../core/utils/parse-boolean.js");
var import_Flex = require("../../../../layouts/flex/Flex.js");
var import_FormFieldContext = require("../../contexts/FormFieldContext.js");
var import_formfield_state_reducer = require("../../state/formfield-state-reducer.js");
const COMPONENT_NAME = "FieldSet";
const RootFieldSet = (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const {
      children,
      disabled: propDisabled,
      name,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "aria-disabled": ariaDisabledProp,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
    const [formFieldState] = (0, import_react.useReducer)(import_formfield_state_reducer._formFieldReducer, {});
    const slots = (0, import_react.useMemo)(() => getSlots(children), [children]);
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_FormFieldContext._FormFieldContext.Provider,
      {
        value: {
          disabled: formFieldState.disabled ?? propDisabled,
          ariaDisabled: formFieldState.ariaDisabled ?? (0, import_parse_boolean.parseBoolean)(ariaDisabledProp),
          required: formFieldState.required,
          dispatch: () => void 0
        },
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          "div",
          {
            "data-testid": dataTestId,
            "data-dtrum-mask": dataDtrumMask,
            "data-dtrum-allow": dataDtrumAllow,
            className: consumerClassName,
            style: consumerStyle,
            ref: forwardedRef,
            ...(0, import_merge_props.mergeProps)(restProps, behavioralTrackingProps),
            "data-dt-component": COMPONENT_NAME,
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("fieldset", { name, className: styles.fieldset, children: [
              slots.legend,
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Flex.Flex, { gap: 12, flexDirection: "column", children: slots.rest })
            ] })
          }
        )
      }
    );
  }
);
RootFieldSet.displayName = COMPONENT_NAME;
function getSlots(children) {
  let legend;
  const rest = [];
  for (const child of import_react.Children.toArray(children)) {
    if ((0, import_react_is.isElement)(child) && child.type === Legend) {
      legend = child;
    } else {
      rest.push(child);
    }
  }
  return { legend, rest };
}
const Legend = (0, import_react.forwardRef)(
  (props, ref) => {
    const { style, className, ...rest } = props;
    const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(rest);
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "legend",
      {
        className: (0, import_clsx.default)(styles.legend, className),
        style,
        ref,
        ...(0, import_merge_props.mergeProps)(restProps, behavioralTrackingProps),
        children: props.children
      }
    );
  }
);
Legend.displayName = "FieldSet.Legend";
const FieldSet = Object.assign(RootFieldSet, {
  Legend
});
