// @dynatrace/strato-components/esm/content/ai-loading-indicator/AiLoadingIndicator.js
import { jsx, jsxs } from "react/jsx-runtime";
import { clsx } from "clsx";
import { useReducedMotion } from "framer-motion";
import { forwardRef } from "react";
import * as styles from "./AiLoadingIndicator.sty.js";
import { AiLoadingIndicatorIcon } from "./AiLoadingIndicatorIcon.js";
import { useAriaLabelingProps } from "../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useId } from "../../core/hooks/useId.js";
import { useSafeForwardProps } from "../../core/hooks/useSafeForwardProps.js";
import { useSlots } from "../../core/hooks/useSlots.js";
import { Flex } from "../../layouts/flex/Flex.js";
const COMPONENT_NAME = "AiLoadingIndicator";
const InternalAiLoadingIndicator = forwardRef((props, forwardedRef) => {
  const {
    id,
    children,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "aria-valuetext": ariaValuetext,
    ...remainingProps
  } = props;
  const [slots, label] = useSlots(children, {
    icon: AiLoadingIndicatorIcon
  });
  const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
  const ariaLabelingProps = useAriaLabelingProps(restProps);
  const forwardProps = useSafeForwardProps(restProps, ariaLabelingProps);
  const { "aria-labelledby": ariaLabelledBy, ...remainingAriaLabelingProps } = ariaLabelingProps;
  const labelId = useId();
  const shouldReduceMotion = useReducedMotion();
  return /* @__PURE__ */ jsxs(
    Flex,
    {
      id,
      ref: forwardedRef,
      role: "progressbar",
      alignItems: "flex-start",
      gap: 8,
      className: clsx(styles.container, consumerClassName),
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "aria-labelledby": ariaLabelledBy ?? labelId,
      "aria-valuetext": ariaValuetext,
      ...remainingAriaLabelingProps,
      ...behavioralTrackingProps,
      ...forwardProps,
      "data-dt-component": COMPONENT_NAME,
      children: [
        slots.icon,
        /* @__PURE__ */ jsx(
          "div",
          {
            id: labelId,
            className: clsx(
              styles.label,
              !shouldReduceMotion && styles.labelAnimation
            ),
            children: label
          }
        )
      ]
    }
  );
});
InternalAiLoadingIndicator.displayName = COMPONENT_NAME;
const AiLoadingIndicator = Object.assign(InternalAiLoadingIndicator, {
  Icon: AiLoadingIndicatorIcon
});
export {
  AiLoadingIndicator
};
