// @dynatrace/strato-components/charts/single-value/SingleValue.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SingleValue_exports = {};
__export(SingleValue_exports, {
  SingleValue: () => SingleValue
});
module.exports = __toCommonJS(SingleValue_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_defaults = require("./defaults.js");
var import_use_single_value_config = require("./hooks/use-single-value-config.js");
var import_SingleValue_css = require("./SingleValue.sty.js");
var import_Sparkline = require("./slots/Sparkline.js");
var import_Toolbar = require("./slots/Toolbar.js");
var import_Trend = require("./slots/Trend.js");
var import_iterate_slots = require("./utils/iterate-slots.js");
var import_single_value_config = require("./utils/single-value-config.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_Intent = require("../../core/slots/Intent/Intent.js");
var import_get_chart_states_template = require("../core/components/chart-states/utils/get-chart-states-template.js");
var import_SingleValueRenderer = require("../core/components/single-value-renderer/SingleValueRenderer.js");
var import_EmptyStateSlot = require("../core/slots/chart-states/EmptyStateSlot.js");
var import_ErrorStateSlot = require("../core/slots/chart-states/ErrorStateSlot.js");
var import_ColorRule = require("../core/slots/ColorRule.js");
var import_chart_types = require("../core/types/chart-types.js");
var import_coerceSizeValue = require("../core/utils/coerceSizeValue.js");
const COMPONENT_NAME = "SingleValue";
const SingleValueRoot = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    data,
    prefixIcon,
    children,
    loading = false,
    width,
    height,
    style: customerStyle,
    className: customerClassNames,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "data-testid": dataTestId,
    "aria-label": ariaLabel,
    "aria-describedby": ariaDescribedBy,
    applyThresholdBackground,
    ...restProps
  } = props;
  const [behavioralTrackingProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(props);
  const { slots, sparklineChildren } = (0, import_iterate_slots.iterateSlots)(children);
  const { emptyState, errorState } = (0, import_get_chart_states_template.getChartStatesTemplates)(children);
  const config = (0, import_use_single_value_config.useSingleValueConfig)(restProps, slots);
  const sparklineProps = slots.sparkline ? {
    data: slots.sparkline?.data ?? [],
    ...config.sparkline
  } : void 0;
  const singleValueRef = (0, import_react.useRef)(null);
  const ref = (0, import_useMergeRefs.useMergeRefs)([
    forwardedRef,
    singleValueRef
  ]);
  (0, import_react.useImperativeHandle)(forwardedRef, () => ({
    getConfig: () => {
      return JSON.stringify({
        type: import_chart_types.ChartType.SingleValue,
        ...(0, import_single_value_config.normalizeConfig)(config)
      });
    }
  }));
  const chartWidth = (0, import_coerceSizeValue.coerceSizeValue)(width);
  const chartHeight = (0, import_coerceSizeValue.coerceSizeValue)(height);
  const mergedStyle = {
    ...customerStyle,
    ...chartWidth && { width: chartWidth },
    ...chartHeight && { height: chartHeight }
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      ...behavioralTrackingProps,
      role: "img",
      className: (0, import_clsx.default)(import_SingleValue_css.container, customerClassNames),
      style: mergedStyle,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "data-testid": dataTestId,
      "data-dt-component": COMPONENT_NAME,
      ref,
      "aria-label": ariaLabel ?? import_defaults.SINGLE_VALUE_ARIA_LABEL,
      "aria-describedby": ariaDescribedBy,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_SingleValueRenderer.SingleValueRenderer,
        {
          data,
          prefixIcon,
          ...config,
          sparklineProps,
          sparklineChildren,
          trendProps: slots.trend,
          emptyState,
          errorState,
          loading,
          applyThresholdBackground,
          children
        }
      )
    }
  );
});
SingleValueRoot.displayName = "SingleValue";
const SingleValue = Object.assign(SingleValueRoot, {
  Sparkline: import_Sparkline.Sparkline,
  Trend: import_Trend.Trend,
  ColorRule: import_ColorRule.ColorRule,
  EmptyState: import_EmptyStateSlot.EmptyStateSlot,
  ErrorState: import_ErrorStateSlot.ErrorStateSlot,
  Intent: import_Intent.Intent,
  Toolbar: import_Toolbar.Toolbar
});
