// @dynatrace/strato-components/charts/xy-chart/processing-strategies/strategies/area-series.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var area_series_exports = {};
__export(area_series_exports, {
  AreaSeriesStrategy: () => AreaSeriesStrategy
});
module.exports = __toCommonJS(area_series_exports);
var import_base_series = require("./base-series.js");
var import_gappable_series = require("./gappable-series.js");
var import_stackable_series = require("./stackable-series.js");
var import_is_valid_points_display = require("../../../core/components/canvas/utils/is-valid-points-display.js");
var import_safe_get = require("../../../core/utils/safe-get.js");
var import_defaults = require("../../constants/defaults.js");
var import_type_guards = require("../../utils/type-guards.js");
var import_get_area_datapoints = require("../utils/access-datapoint/get-area-datapoints.js");
var import_get_datapoints = require("../utils/access-datapoint/get-datapoints.js");
var import_get_calculated_linear_point = require("../utils/cast-data/get-calculated-linear-point.js");
var import_get_linear_color_palette = require("../utils/coloring/get-linear-color-palette.js");
var import_is_area_series_stacked = require("../utils/stacking/is-area-series-stacked.js");
var import_stack_datapoints = require("../utils/stacking/stack-datapoints.js");
class AreaSeriesStrategy extends import_base_series.SeriesBaseStrategy {
  stackHelper;
  gapHelper;
  /**
   * This is needed for special cases such as isolated points in value representation in area charts.
   */
  combinedAxesOrientationKeyToSeriesCount;
  /** This set contains series Id that has been preprocessed at least once */
  preprocessedSeriesIds;
  constructor(initializer) {
    super(initializer);
    this.stackHelper = new import_stackable_series.StackableSeriesHelper();
    this.gapHelper = new import_gappable_series.GappableSeriesHelper();
    this.combinedAxesOrientationKeyToSeriesCount = /* @__PURE__ */ new Map();
    this.preprocessedSeriesIds = /* @__PURE__ */ new Set();
  }
  canHandle(seriesProps) {
    return (0, import_type_guards.isAreaSeries)(seriesProps);
  }
  getSeriesId(datum, seriesProps) {
    return (0, import_get_datapoints.getDatumValue)(
      datum,
      seriesProps.seriesIdAccessor
    )?.toString();
  }
  createSeries(seriesId, seriesProps, xAxis, yAxis) {
    return {
      id: seriesId,
      type: seriesProps.type,
      data: [],
      xAxisId: xAxis.id,
      yAxisId: yAxis.id
    };
  }
  createSeriesMetadata(datum, seriesProps, xAxis, yAxis, seriesId, context) {
    const { intl, seriesIndexes } = context;
    const { nameAccessor: seriesNameAccessor, curve } = seriesProps;
    const seriesIndex = this.getOrCreateSeriesIndex(seriesId, seriesIndexes);
    const seriesName = this.getSeriesName(intl, seriesIndex, void 0, {
      seriesNameAccessor,
      datum
    });
    const color = this.getColor(
      datum,
      seriesProps,
      context.colorPalette,
      seriesName,
      seriesIndex
    );
    const chartOrientation = (0, import_get_datapoints.getChartOrientation)(seriesProps);
    const orientation = chartOrientation === "horizontal" ? "horizontal" : "vertical";
    const variant = (0, import_get_datapoints.getChartAreaVariant)(
      chartOrientation,
      seriesProps
    );
    return {
      xAxis,
      yAxis,
      ...color,
      name: seriesName,
      orientation,
      curve,
      variant,
      pointsDisplay: seriesProps.pointsDisplay && (0, import_is_valid_points_display.isValidPointsDisplay)(seriesProps.pointsDisplay) ? seriesProps.pointsDisplay : import_defaults.DEFAULT_POINTS_DISPLAY,
      actions: seriesProps.actions
    };
  }
  getCalculationPoint(datapoint, xAxis, yAxis, chartOrientation, variant) {
    return (0, import_get_calculated_linear_point.getCalculatedLinearPointForArea)(
      datapoint,
      xAxis,
      yAxis,
      chartOrientation,
      variant
    );
  }
  preprocessImpl(data, seriesProps, xAxis, yAxis, context) {
    const chartOrientation = (0, import_get_datapoints.getChartOrientation)(seriesProps);
    if (chartOrientation === void 0 || chartOrientation === "both") {
      return;
    }
    const areaVariant = (0, import_get_datapoints.getChartAreaVariant)(chartOrientation, seriesProps);
    if (areaVariant === "band") {
      return;
    }
    const canStack = (0, import_stack_datapoints.isStackable)("stack", chartOrientation, xAxis, yAxis);
    if (!canStack) {
      return;
    }
    const filteredSeriesIds = context?.filteredSeriesIds;
    for (const datum of data) {
      const datapoint = this.getDatapoint(datum, seriesProps, xAxis, yAxis);
      const variant = (0, import_get_datapoints.getChartAreaVariant)(chartOrientation, seriesProps);
      if (datapoint) {
        const seriesId = this.getSeriesId(datum, seriesProps);
        const isSeriesVisible = filteredSeriesIds === void 0 || filteredSeriesIds.has(seriesId);
        if (!isSeriesVisible) {
          continue;
        }
        this.registerSeriesCountOnPreprocessing(
          seriesId,
          xAxis.id,
          yAxis.id,
          chartOrientation
        );
        const preCalcPoint = this.getCalculationPoint(
          datapoint,
          xAxis,
          yAxis,
          chartOrientation,
          variant
        );
        this.stackHelper.processStackDomain(
          preCalcPoint,
          chartOrientation,
          xAxis,
          yAxis,
          seriesId
        );
      }
    }
  }
  processImpl(datum, seriesProps, xAxis, yAxis, datapoint, context) {
    const { seriesMap, seriesToMetadata } = context;
    const seriesId = this.getSeriesId(datum, seriesProps);
    const isSeriesVisible = context.filteredSeriesIds === void 0 || context.filteredSeriesIds.has(seriesId);
    const chartOrientation = (0, import_get_datapoints.getChartOrientation)(seriesProps);
    if (!chartOrientation) {
      return "continue";
    }
    const variant = (0, import_get_datapoints.getChartAreaVariant)(chartOrientation, seriesProps);
    const preCalcPoint = this.getCalculationPoint(
      datapoint,
      xAxis,
      yAxis,
      chartOrientation,
      variant
    );
    let calcDatapoint;
    let datapointMetadata;
    let shouldContinue;
    if (isSeriesVisible) {
      ({ calcDatapoint, datapointMetadata, shouldContinue } = this.handleStacking(
        preCalcPoint,
        xAxis,
        yAxis,
        chartOrientation,
        variant
      ));
    } else {
      calcDatapoint = preCalcPoint;
      datapointMetadata = {
        isGap: false,
        baseline: (0, import_stack_datapoints.getBaselineCoordinate)(chartOrientation)
      };
      shouldContinue = false;
    }
    if (shouldContinue) {
      return "continue";
    }
    let currentSeries = seriesMap.get(seriesId);
    if (!currentSeries) {
      currentSeries = this.createSeries(seriesId, seriesProps, xAxis, yAxis);
      this.registerSeries(currentSeries, seriesId, context);
      const metadata = this.createSeriesMetadata(
        datum,
        seriesProps,
        xAxis,
        yAxis,
        seriesId,
        context
      );
      seriesToMetadata.set(currentSeries, {
        ...metadata,
        isVisible: isSeriesVisible,
        isStacked: chartOrientation !== "both" && (0, import_is_area_series_stacked.isAreaSeriesStacked)(
          (0, import_stack_datapoints.getCombinedAxesOrientationKey)(xAxis.id, yAxis.id, chartOrientation),
          this.combinedAxesOrientationKeyToSeriesCount,
          variant
        )
      });
    }
    const currentSeriesMetadata = (0, import_safe_get.safeGet)(seriesToMetadata, currentSeries);
    if (isSeriesVisible) {
      context.axisWithSeries.add(xAxis.id);
      context.axisWithSeries.add(yAxis.id);
      const gappedDatapoint = this.gapHelper.handleGapDetection(
        datapoint,
        currentSeries,
        xAxis,
        yAxis
      );
      if (gappedDatapoint) {
        this.registerDatapoint(
          gappedDatapoint,
          { isGap: true },
          currentSeries,
          context
        );
      }
    }
    this.registerDatapoint(
      datapoint,
      datapointMetadata,
      currentSeries,
      context
    );
    this.updateAxisDomains(
      calcDatapoint,
      xAxis,
      yAxis,
      context.axisIdToScaleDomain,
      isSeriesVisible
    );
    if (isSeriesVisible) {
      this.addValue({
        series: currentSeries,
        datapoint,
        color: currentSeriesMetadata?.color
      });
    }
  }
  getColor(datum, seriesProps, configColorPalette, seriesName, seriesIndex) {
    return (0, import_get_linear_color_palette.getLinearColor)(
      datum,
      seriesProps,
      configColorPalette,
      seriesName,
      seriesIndex
    );
  }
  getDatapoint(datum, seriesProps, xAxis, yAxis) {
    return (0, import_get_area_datapoints.getAreaDatapoints)(datum, seriesProps, xAxis, yAxis);
  }
  handleStacking(datapoint, xAxis, yAxis, chartOrientation, areaVariant) {
    if (areaVariant === "band") {
      return {
        calcDatapoint: datapoint,
        datapointMetadata: {
          isGap: false,
          baseline: (0, import_stack_datapoints.getBaselineCoordinate)(chartOrientation)
        },
        shouldContinue: false
      };
    }
    let calcDatapoint;
    let shouldContinue = false;
    let datapointMetadata = {
      isGap: false,
      baseline: (0, import_stack_datapoints.getBaselineCoordinate)(chartOrientation)
    };
    const canStack = (0, import_stack_datapoints.isStackable)("stack", chartOrientation, xAxis, yAxis);
    if (canStack) {
      ({ calcDatapoint, datapointMetadata, shouldContinue } = this.stackHelper.processStackedDatapoint(
        datapoint,
        chartOrientation,
        xAxis,
        yAxis,
        datapointMetadata
      ));
    }
    return {
      calcDatapoint: calcDatapoint ?? datapoint,
      datapointMetadata,
      shouldContinue
    };
  }
  registerSeriesCountForAxisCombination(xAxisId, yAxisId, chartOrientation) {
    const axesCombinationKey = (0, import_stack_datapoints.getCombinedAxesOrientationKey)(
      xAxisId,
      yAxisId,
      chartOrientation
    );
    const count = this.combinedAxesOrientationKeyToSeriesCount.get(axesCombinationKey) ?? 0;
    this.combinedAxesOrientationKeyToSeriesCount.set(
      axesCombinationKey,
      count + 1
    );
  }
  registerSeriesCountOnPreprocessing(seriesId, xAxisId, yAxisId, chartOrientation) {
    if (!this.preprocessedSeriesIds.has(seriesId)) {
      this.registerSeriesCountForAxisCombination(
        xAxisId,
        yAxisId,
        chartOrientation
      );
      this.preprocessedSeriesIds.add(seriesId);
    }
  }
}
