// @dynatrace/strato-components/charts/xy-chart/processing-strategies/strategies/bar-series.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var bar_series_exports = {};
__export(bar_series_exports, {
  BarSeriesStrategy: () => BarSeriesStrategy
});
module.exports = __toCommonJS(bar_series_exports);
var import_base_series = require("./base-series.js");
var import_groupable_series = require("./groupable-series.js");
var import_stackable_series = require("./stackable-series.js");
var import_xy_chart = require("../../types/xy-chart.js");
var import_type_guards = require("../../utils/type-guards.js");
var import_get_bar_datapoints = require("../utils/access-datapoint/get-bar-datapoints.js");
var import_get_datapoints = require("../utils/access-datapoint/get-datapoints.js");
var import_get_linear_color_palette = require("../utils/coloring/get-linear-color-palette.js");
var import_get_variant = require("../utils/grouping/get-variant.js");
var import_group_bar_datapoints = require("../utils/grouping/group-bar-datapoints.js");
var import_stack_datapoints = require("../utils/stacking/stack-datapoints.js");
class BarSeriesStrategy extends import_base_series.SeriesBaseStrategy {
  stackHelper;
  groupHelper;
  constructor(initializer) {
    super(initializer);
    this.stackHelper = new import_stackable_series.StackableSeriesHelper();
    this.groupHelper = new import_groupable_series.GroupableSeriesHelper();
  }
  get axisIdToDimension() {
    return this.groupHelper.axisIdToDimension;
  }
  canHandle(seriesProps) {
    return (0, import_type_guards.isBarSeries)(seriesProps);
  }
  getSeriesId(datum, seriesProps) {
    return (0, import_get_datapoints.getDatumValue)(
      datum,
      seriesProps.seriesIdAccessor
    )?.toString();
  }
  createSeries(seriesId, seriesProps, xAxis, yAxis) {
    return {
      id: seriesId,
      type: seriesProps.type,
      data: [],
      xAxisId: xAxis.id,
      yAxisId: yAxis.id,
      labelAccessor: seriesProps.labelAccessor
    };
  }
  createSeriesMetadata(datum, seriesProps, xAxis, yAxis, seriesId, context, variant) {
    const { intl, seriesIndexes } = context;
    const { nameAccessor: seriesNameAccessor } = seriesProps;
    const seriesIndex = this.getOrCreateSeriesIndex(seriesId, seriesIndexes);
    const seriesName = this.getSeriesName(intl, seriesIndex, void 0, {
      seriesNameAccessor,
      datum
    });
    const color = this.getColor(
      datum,
      seriesProps,
      context.colorPalette,
      seriesName,
      seriesIndex
    );
    return {
      xAxis,
      yAxis,
      ...color,
      name: seriesName,
      variant: variant ?? import_xy_chart.BAR_VARIANT_SINGLE,
      actions: seriesProps.actions,
      chartOrientation: (0, import_get_datapoints.getChartOrientation)(seriesProps)
    };
  }
  handleStackingAndGrouping(datapoint, seriesProps, xAxis, yAxis, chartOrientation, seriesId, isSeriesVisible) {
    let shouldContinue = false;
    let datapointMetadata = {
      isGap: false,
      baseline: (0, import_stack_datapoints.getBaselineCoordinate)(chartOrientation)
    };
    const canStack = (0, import_stack_datapoints.isStackable)(
      seriesProps.variant,
      chartOrientation,
      xAxis,
      yAxis
    );
    const canGroup = (0, import_group_bar_datapoints.isGrouping)(
      seriesProps.variant,
      chartOrientation,
      xAxis,
      yAxis
    );
    let calcDatapoint;
    if (!isSeriesVisible) {
      return {
        calcDatapoint: calcDatapoint ?? datapoint,
        datapointMetadata,
        shouldContinue,
        variant: (0, import_get_variant.getBarSeriesVariant)(canStack, canGroup)
      };
    }
    if (canStack) {
      ({ calcDatapoint, datapointMetadata, shouldContinue } = this.stackHelper.processStackedDatapoint(
        datapoint,
        chartOrientation,
        xAxis,
        yAxis,
        datapointMetadata
      ));
    }
    if (canGroup) {
      ({ datapointMetadata } = this.groupHelper.processGroupDatapoint(
        datapoint,
        chartOrientation,
        xAxis,
        yAxis,
        seriesId,
        datapointMetadata
      ));
    }
    return {
      calcDatapoint: calcDatapoint ?? datapoint,
      datapointMetadata,
      shouldContinue,
      variant: (0, import_get_variant.getBarSeriesVariant)(canStack, canGroup)
    };
  }
  preprocessImpl(data, seriesProps, xAxis, yAxis, context) {
    const chartOrientation = (0, import_get_datapoints.getChartOrientation)(seriesProps);
    if (chartOrientation === void 0 || chartOrientation === "both") {
      return;
    }
    const canStack = (0, import_stack_datapoints.isStackable)(
      seriesProps.variant,
      chartOrientation,
      xAxis,
      yAxis
    );
    const canGroup = (0, import_group_bar_datapoints.isGrouping)(
      seriesProps.variant,
      chartOrientation,
      xAxis,
      yAxis
    );
    if (!canStack && !canGroup) {
      return;
    }
    const filteredSeriesIds = context?.filteredSeriesIds;
    for (const datum of data) {
      const datapoint = this.getDatapoint(datum, seriesProps, xAxis, yAxis);
      if (datapoint) {
        const seriesId = this.getSeriesId(datum, seriesProps);
        const isSeriesVisible = filteredSeriesIds === void 0 || filteredSeriesIds.has(seriesId);
        if (!isSeriesVisible) {
          continue;
        }
        if (canStack) {
          this.stackHelper.processStackDomain(
            datapoint,
            chartOrientation,
            xAxis,
            yAxis,
            seriesId
          );
        }
        if (canGroup) {
          this.groupHelper.processGroupDomain(
            datapoint,
            chartOrientation,
            xAxis,
            yAxis,
            seriesId
          );
        }
      }
    }
  }
  processImpl(datum, seriesProps, xAxis, yAxis, datapoint, context) {
    const { seriesMap, seriesToMetadata } = context;
    const seriesId = this.getSeriesId(datum, seriesProps);
    const chartOrientation = (0, import_get_datapoints.getChartOrientation)(seriesProps);
    if (chartOrientation === void 0) {
      return "continue";
    }
    const isSeriesVisible = context.filteredSeriesIds === void 0 || context.filteredSeriesIds.has(seriesId);
    const { calcDatapoint, datapointMetadata, shouldContinue, variant } = this.handleStackingAndGrouping(
      datapoint,
      seriesProps,
      xAxis,
      yAxis,
      chartOrientation,
      seriesId,
      isSeriesVisible
    );
    if (shouldContinue) {
      return "continue";
    }
    let currentSeries = seriesMap.get(seriesId);
    if (!currentSeries) {
      currentSeries = this.createSeries(seriesId, seriesProps, xAxis, yAxis);
      this.registerSeries(currentSeries, seriesId, context);
      const metadata = this.createSeriesMetadata(
        datum,
        seriesProps,
        xAxis,
        yAxis,
        seriesId,
        context,
        variant
      );
      seriesToMetadata.set(currentSeries, {
        ...metadata,
        isVisible: isSeriesVisible
      });
    }
    if (datapoint.label) {
      currentSeries.hasLabels = true;
    }
    this.registerDatapoint(
      datapoint,
      datapointMetadata,
      currentSeries,
      context
    );
    this.updateAxisDomains(
      calcDatapoint,
      xAxis,
      yAxis,
      context.axisIdToScaleDomain,
      isSeriesVisible
    );
    if (isSeriesVisible) {
      context.axisWithSeries.add(xAxis.id);
      context.axisWithSeries.add(yAxis.id);
      this.addValue({
        series: currentSeries,
        datapoint
      });
    }
  }
  getColor(datum, seriesProps, configColorPalette, seriesName, seriesIndex) {
    return (0, import_get_linear_color_palette.getLinearColor)(
      datum,
      seriesProps,
      configColorPalette,
      seriesName,
      seriesIndex
    );
  }
  getDatapoint(datum, seriesProps, xAxis, yAxis) {
    return (0, import_get_bar_datapoints.getBarDatapoint)(datum, seriesProps, xAxis, yAxis);
  }
}
