// @dynatrace/strato-components/charts/xy-chart/processing-strategies/strategies/base-series.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var base_series_exports = {};
__export(base_series_exports, {
  SeriesBaseStrategy: () => SeriesBaseStrategy
});
module.exports = __toCommonJS(base_series_exports);
var import_lodash_es = require("lodash");
var import_axis = require("../../types/axis.js");
var import_calculate_axes_domains = require("../../utils/calculate-axes-domains.js");
var import_get_series_name = require("../utils/series-naming/get-series-name.js");
class SeriesBaseStrategy {
  addValue;
  constructor(initializer) {
    this.addValue = initializer.addValue;
  }
  preprocess(data, seriesProps, xAxis, yAxis, context) {
    if (!this.canHandle(seriesProps)) {
      return;
    }
    return this.preprocessImpl(data, seriesProps, xAxis, yAxis, context);
  }
  process(datum, seriesProps, xAxis, yAxis, context) {
    if (!this.canHandle(seriesProps)) {
      return;
    }
    const datapoint = this.getDatapoint(datum, seriesProps, xAxis, yAxis);
    if (!datapoint) {
      return "continue";
    }
    return this.processImpl(
      datum,
      seriesProps,
      xAxis,
      yAxis,
      datapoint,
      context
    );
  }
  preprocessImpl(data, seriesProps, xAxis, yAxis, context) {
    return;
  }
  /**
   * Gets or creates the series index, adding to seriesIndexes if new.
   */
  getOrCreateSeriesIndex(seriesId, seriesIndexes) {
    if (!seriesIndexes.includes(seriesId)) {
      seriesIndexes.push(seriesId);
    }
    return seriesIndexes.indexOf(seriesId);
  }
  /**
   * Creates a new series object. Override in subclasses for custom series structure.
   */
  createSeries(seriesId, seriesProps, xAxis, yAxis) {
    return {
      id: seriesId,
      type: seriesProps.type,
      data: [],
      xAxisId: xAxis.id,
      yAxisId: yAxis.id
    };
  }
  /**
   * Registers a new series in the context maps.
   */
  registerSeries(currentSeries, seriesId, context) {
    context.seriesMap.set(seriesId, currentSeries);
    context.seriesWithDatapoints.push(currentSeries);
  }
  /**
   * Registers a datapoint to the series and updates metadata and domains.
   */
  registerDatapoint(datapoint, datapointMetadata, currentSeries, context) {
    currentSeries.data.push(datapoint);
    context.datapointMetadataMap.set(datapoint, datapointMetadata);
  }
  /**
   * Calculates and sets the axis domains.
   */
  updateAxisDomains(calcPoint, xAxis, yAxis, axisIdToScaleDomain, isSeriesVisible) {
    const { xDomain, yDomain } = (0, import_calculate_axes_domains.calculateAxesDomains)(
      xAxis,
      yAxis,
      calcPoint,
      axisIdToScaleDomain,
      isSeriesVisible
    );
    axisIdToScaleDomain.set(xAxis.id, xDomain);
    if (isSeriesVisible || yAxis.type === import_axis.XYAxisTypeCategorical) {
      axisIdToScaleDomain.set(yAxis.id, yDomain);
    }
  }
  /**
   * Gets the calculation point for domain updates. Override for custom logic (e.g., time centering).
   */
  getCalculationPoint(datapoint, _xAxis, _yAxis) {
    return datapoint;
  }
  getSeriesName(intl, seriesIndex, seriesNameFallback, namingWithAccessor) {
    return (0, import_lodash_es.isUndefined)(namingWithAccessor) ? (0, import_get_series_name.getSeriesNameWithFallback)(intl, seriesIndex, seriesNameFallback) : (0, import_get_series_name.getSeriesNameWithAccessor)(
      intl,
      seriesIndex,
      seriesNameFallback,
      namingWithAccessor.seriesNameAccessor,
      namingWithAccessor.datum
    );
  }
}
