// @dynatrace/strato-components/charts/xy-chart/processing-strategies/strategies/dot-series.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var dot_series_exports = {};
__export(dot_series_exports, {
  DotSeriesStrategy: () => DotSeriesStrategy
});
module.exports = __toCommonJS(dot_series_exports);
var import_base_series = require("./base-series.js");
var import_safe_get = require("../../../core/utils/safe-get.js");
var import_type_guards = require("../../utils/type-guards.js");
var import_get_datapoints = require("../utils/access-datapoint/get-datapoints.js");
var import_get_calculated_linear_point = require("../utils/cast-data/get-calculated-linear-point.js");
var import_get_linear_color_palette = require("../utils/coloring/get-linear-color-palette.js");
class DotSeriesStrategy extends import_base_series.SeriesBaseStrategy {
  canHandle(seriesProps) {
    return (0, import_type_guards.isDotSeries)(seriesProps);
  }
  // can this method be un base?
  getSeriesId(datum, seriesProps) {
    const seriesIdAccessor = (0, import_get_datapoints.getDatumValue)(
      datum,
      seriesProps.seriesIdAccessor
    )?.toString();
    return seriesIdAccessor ?? seriesProps.id;
  }
  createSeriesMetadata(datum, seriesProps, xAxis, yAxis, seriesId, context) {
    const { intl, seriesIndexes } = context;
    const { nameAccessor: seriesNameAccessor } = seriesProps;
    const seriesIndex = this.getOrCreateSeriesIndex(seriesId, seriesIndexes);
    const seriesName = this.getSeriesName(
      intl,
      seriesIndex,
      // APPDEV-15290 - Adjust after removing name prop from series
      // eslint-disable-next-line @typescript-eslint/no-deprecated
      seriesProps.name,
      {
        seriesNameAccessor,
        datum
      }
    );
    const color = this.getColor(
      datum,
      seriesProps,
      context.colorPalette,
      seriesName,
      seriesIndex
    );
    return {
      xAxis,
      yAxis,
      actions: seriesProps.actions,
      ...color,
      name: seriesName,
      shape: seriesProps.shape
    };
  }
  getCalculationPoint(datapoint, xAxis, yAxis) {
    return (0, import_get_calculated_linear_point.getCalculatedLinearPoint)(datapoint, xAxis, yAxis);
  }
  processImpl(datum, seriesProps, xAxis, yAxis, datapoint, context) {
    const { seriesMap, seriesToMetadata } = context;
    const seriesId = this.getSeriesId(datum, seriesProps);
    const isSeriesVisible = context.filteredSeriesIds === void 0 || context.filteredSeriesIds.has(seriesId);
    let currentSeries = seriesMap.get(seriesId);
    if (!currentSeries) {
      currentSeries = this.createSeries(
        seriesId,
        seriesProps,
        xAxis,
        yAxis
      );
      this.registerSeries(currentSeries, seriesId, context);
      const metadata = this.createSeriesMetadata(
        datum,
        seriesProps,
        xAxis,
        yAxis,
        seriesId,
        context
      );
      seriesToMetadata.set(currentSeries, {
        ...metadata,
        isVisible: isSeriesVisible
      });
    }
    const datapointMetadata = { isGap: false };
    this.registerDatapoint(
      datapoint,
      datapointMetadata,
      currentSeries,
      context
    );
    const calcPoint = this.getCalculationPoint(datapoint, xAxis, yAxis);
    this.updateAxisDomains(
      calcPoint,
      xAxis,
      yAxis,
      context.axisIdToScaleDomain,
      isSeriesVisible
    );
    if (isSeriesVisible) {
      context.axisWithSeries.add(xAxis.id);
      context.axisWithSeries.add(yAxis.id);
      const currentSeriesMetadata = (0, import_safe_get.safeGet)(seriesToMetadata, currentSeries);
      this.addValue({
        series: currentSeries,
        datapoint,
        color: currentSeriesMetadata?.color
      });
    }
  }
  getColor(datum, seriesProps, configColorPalette, seriesName, seriesIndex) {
    return (0, import_get_linear_color_palette.getLinearColor)(
      datum,
      seriesProps,
      configColorPalette,
      seriesName,
      seriesIndex
    );
  }
  getDatapoint(datum, seriesProps, xAxis, yAxis) {
    return (0, import_get_datapoints.getDotDatapoint)(datum, seriesProps, xAxis, yAxis);
  }
}
