// @dynatrace/strato-components/charts/xy-chart/processing-strategies/strategies/groupable-series.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var groupable_series_exports = {};
__export(groupable_series_exports, {
  GroupableSeriesHelper: () => GroupableSeriesHelper
});
module.exports = __toCommonJS(groupable_series_exports);
var import_stack_datapoints = require("../utils/stacking/stack-datapoints.js");
class GroupableSeriesHelper {
  static STARTING_POINT = 0;
  groupData;
  axisStats;
  axisIdToDimension;
  constructor() {
    this.groupData = /* @__PURE__ */ new Map();
    this.axisStats = /* @__PURE__ */ new Map();
    this.axisIdToDimension = /* @__PURE__ */ new Map();
  }
  /**
   * Calculate total height of group bin.
   * Should be called at the beginning of a new grouping calculation cycle (preprocess phase).
   *
   * Also records which series contributed to each axisKey so that `getGroupBinDomain` can
   * fall back to a cross-bin reference for single-series axes. Single-series groups would
   * otherwise collapse to 100% in every bin.
   */
  processGroupDomain(datapoint, chartOrientation, xAxis, yAxis, seriesId) {
    const { stackKey, value } = (0, import_stack_datapoints.getStackingData)(
      chartOrientation,
      datapoint,
      xAxis,
      yAxis
    );
    const axisKey = (0, import_stack_datapoints.getCombinedAxesOrientationKey)(
      xAxis.id,
      yAxis.id,
      chartOrientation
    );
    const stats = this.getOrInitAxisStats(axisKey);
    stats.seriesIds.add(seriesId);
    if (typeof value === "number") {
      const acc = this.groupData.get(stackKey) ?? {
        totalPositive: GroupableSeriesHelper.STARTING_POINT,
        totalNegative: GroupableSeriesHelper.STARTING_POINT,
        axisKey
      };
      if (value >= 0) {
        acc.totalPositive += value;
      } else {
        acc.totalNegative += Math.abs(value);
      }
      this.groupData.set(stackKey, acc);
      const binAbsValue = Math.max(acc.totalPositive, acc.totalNegative);
      if (binAbsValue > stats.maxBinAbsValue) {
        stats.maxBinAbsValue = binAbsValue;
      }
    }
  }
  /**
   * Returns the total bin domain for a given group key.
   *
   * For axes with more than one contributing series the result is the per-bin total height
   * of the group (|positive| + |negative|). For axes with only one contributing series the
   * result collapses to a cross-bin reference: the largest per-bin absolute total across all
   * bins of that axis. This keeps single-series groups consistent with the regular relative
   * behavior, where the tallest bar represents 100% and the rest scale proportionally.
   */
  getGroupBinDomain(groupKey) {
    const groupDomainObj = this.groupData.get(groupKey);
    if (!groupDomainObj) {
      return void 0;
    }
    const stats = this.axisStats.get(groupDomainObj.axisKey);
    if (stats && stats.seriesIds.size === 1) {
      return stats.maxBinAbsValue;
    }
    return groupDomainObj.totalPositive + groupDomainObj.totalNegative;
  }
  getOrInitAxisStats(axisKey) {
    let stats = this.axisStats.get(axisKey);
    if (!stats) {
      stats = { seriesIds: /* @__PURE__ */ new Set(), maxBinAbsValue: 0 };
      this.axisStats.set(axisKey, stats);
    }
    return stats;
  }
  /**
   * Processes a datapoint for grouping, calculating the bin domain metadata.
   * Also registers the series in the dimension map for the primary axis.
   * Returns metadata with binDomain set but no accumulator (bars remain side-by-side).
   */
  processGroupDatapoint(datapoint, chartOrientation, xAxis, yAxis, seriesId, datapointMetadata) {
    const primaryAxis = chartOrientation === "vertical" ? xAxis : yAxis;
    const dimension = this.axisIdToDimension.get(primaryAxis.id);
    if (dimension) {
      dimension.add(seriesId);
      this.axisIdToDimension.set(primaryAxis.id, dimension);
    } else {
      this.axisIdToDimension.set(
        primaryAxis.id,
        (/* @__PURE__ */ new Set()).add(seriesId)
      );
    }
    const { stackKey, value } = (0, import_stack_datapoints.getStackingData)(
      chartOrientation,
      datapoint,
      xAxis,
      yAxis
    );
    if (typeof value !== "number") {
      return { datapointMetadata };
    }
    const binDomain = this.getGroupBinDomain(stackKey);
    return {
      datapointMetadata: {
        ...datapointMetadata,
        binDomain
      }
    };
  }
}
