// @dynatrace/strato-components/charts/xy-chart/processing-strategies/strategies/line-series.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var line_series_exports = {};
__export(line_series_exports, {
  LineSeriesStrategy: () => LineSeriesStrategy
});
module.exports = __toCommonJS(line_series_exports);
var import_base_series = require("./base-series.js");
var import_gappable_series = require("./gappable-series.js");
var import_is_valid_points_display = require("../../../core/components/canvas/utils/is-valid-points-display.js");
var import_safe_get = require("../../../core/utils/safe-get.js");
var import_defaults = require("../../constants/defaults.js");
var import_type_guards = require("../../utils/type-guards.js");
var import_get_datapoints = require("../utils/access-datapoint/get-datapoints.js");
var import_get_calculated_linear_point = require("../utils/cast-data/get-calculated-linear-point.js");
var import_get_linear_color_palette = require("../utils/coloring/get-linear-color-palette.js");
class LineSeriesStrategy extends import_base_series.SeriesBaseStrategy {
  gapHelper;
  constructor(initializer) {
    super(initializer);
    this.gapHelper = new import_gappable_series.GappableSeriesHelper();
  }
  canHandle(seriesProps) {
    return (0, import_type_guards.isLineSeries)(seriesProps);
  }
  getSeriesId(datum, seriesProps) {
    return (0, import_get_datapoints.getDatumValue)(
      datum,
      seriesProps.seriesIdAccessor
    )?.toString();
  }
  createSeriesMetadata(datum, seriesProps, xAxis, yAxis, seriesId, context) {
    const { intl, seriesIndexes } = context;
    const { nameAccessor: seriesNameAccessor } = seriesProps;
    const seriesIndex = this.getOrCreateSeriesIndex(seriesId, seriesIndexes);
    const seriesName = this.getSeriesName(intl, seriesIndex, void 0, {
      seriesNameAccessor,
      datum
    });
    const color = this.getColor(
      datum,
      seriesProps,
      context.colorPalette,
      seriesName,
      seriesIndex
    );
    return {
      xAxis,
      yAxis,
      curve: seriesProps.curve,
      actions: seriesProps.actions,
      pointsDisplay: seriesProps.pointsDisplay && (0, import_is_valid_points_display.isValidPointsDisplay)(seriesProps.pointsDisplay) ? seriesProps.pointsDisplay : import_defaults.DEFAULT_POINTS_DISPLAY,
      ...color,
      name: seriesName
    };
  }
  getCalculationPoint(datapoint, xAxis, yAxis) {
    return (0, import_get_calculated_linear_point.getCalculatedLinearPoint)(datapoint, xAxis, yAxis);
  }
  processImpl(datum, seriesProps, xAxis, yAxis, datapoint, context) {
    const { seriesMap, seriesToMetadata } = context;
    const seriesId = this.getSeriesId(datum, seriesProps);
    const isSeriesVisible = context.filteredSeriesIds === void 0 || context.filteredSeriesIds.has(seriesId);
    let currentSeries = seriesMap.get(seriesId);
    if (!currentSeries) {
      currentSeries = this.createSeries(
        seriesId,
        seriesProps,
        xAxis,
        yAxis
      );
      this.registerSeries(currentSeries, seriesId, context);
      const metadata = this.createSeriesMetadata(
        datum,
        seriesProps,
        xAxis,
        yAxis,
        seriesId,
        context
      );
      seriesToMetadata.set(currentSeries, {
        ...metadata,
        isVisible: isSeriesVisible
      });
    }
    const datapointMetadata = { isGap: false };
    if (isSeriesVisible) {
      const gappedDatapoint = this.gapHelper.handleGapDetection(
        datapoint,
        currentSeries,
        xAxis,
        yAxis
      );
      if (gappedDatapoint) {
        this.registerDatapoint(
          gappedDatapoint,
          { isGap: true },
          currentSeries,
          context
        );
      }
    }
    this.registerDatapoint(
      datapoint,
      datapointMetadata,
      currentSeries,
      context
    );
    const calcPoint = this.getCalculationPoint(datapoint, xAxis, yAxis);
    this.updateAxisDomains(
      calcPoint,
      xAxis,
      yAxis,
      context.axisIdToScaleDomain,
      isSeriesVisible
    );
    if (isSeriesVisible) {
      context.axisWithSeries.add(xAxis.id);
      context.axisWithSeries.add(yAxis.id);
      const currentSeriesMetadata = (0, import_safe_get.safeGet)(seriesToMetadata, currentSeries);
      this.addValue({
        series: currentSeries,
        datapoint,
        color: currentSeriesMetadata?.color
      });
    }
  }
  getColor(datum, seriesProps, configColorPalette, seriesName, seriesIndex) {
    return (0, import_get_linear_color_palette.getLinearColor)(
      datum,
      seriesProps,
      configColorPalette,
      seriesName,
      seriesIndex
    );
  }
  getDatapoint(datum, seriesProps, xAxis, yAxis) {
    return (0, import_get_datapoints.getLineDatapoint)(datum, seriesProps, xAxis, yAxis);
  }
}
