// @dynatrace/strato-components/charts/xy-chart/processing-strategies/strategies/rect-series.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var rect_series_exports = {};
__export(rect_series_exports, {
  RectSeriesStrategy: () => RectSeriesStrategy
});
module.exports = __toCommonJS(rect_series_exports);
var import_base_series = require("./base-series.js");
var import_safe_get = require("../../../core/utils/safe-get.js");
var import_defaults = require("../../constants/defaults.js");
var import_xy_chart_internals = require("../../types/xy-chart-internals.js");
var import_type_guards = require("../../utils/type-guards.js");
var import_get_datapoints = require("../utils/access-datapoint/get-datapoints.js");
var import_get_legend_type = require("../utils/cast-data/get-legend-type.js");
var import_get_valid_color_palette = require("../utils/coloring/get-valid-color-palette.js");
var import_get_value_scale_domain = require("../utils/scale/get-value-scale-domain.js");
class RectSeriesStrategy extends import_base_series.SeriesBaseStrategy {
  seriesIdToValueScaleDomain;
  constructor(initializer) {
    super(initializer);
    this.seriesIdToValueScaleDomain = /* @__PURE__ */ new Map();
  }
  canHandle(seriesProps) {
    return (0, import_type_guards.isRectSeries)(seriesProps);
  }
  getSeriesId(_datum, seriesProps) {
    return seriesProps.id;
  }
  createSeries(seriesId, seriesProps, xAxis, yAxis) {
    return {
      id: seriesId,
      type: seriesProps.type,
      data: [],
      filteredData: [],
      xAxisId: xAxis.id,
      yAxisId: yAxis.id
    };
  }
  createSeriesMetadata(_datum, seriesProps, xAxis, yAxis, seriesId, context) {
    const { intl, seriesIndexes } = context;
    const color = this.getColor(seriesProps.colorPalette, context.colorPalette);
    const seriesIndex = this.getOrCreateSeriesIndex(seriesId, seriesIndexes);
    const seriesName = this.getSeriesName(intl, seriesIndex, void 0);
    return {
      name: seriesName,
      xAxis,
      yAxis,
      actions: seriesProps.actions,
      valueFormatter: seriesProps.valueFormatter,
      valueUnit: seriesProps.valueUnit,
      colorPalette: color.colorPalette,
      valueAccessor: seriesProps.valueAccessor,
      colorIsFromSlot: color.colorIsFromSlot,
      valueType: void 0,
      valueAccessorLabel: seriesProps.valueAccessorLabel,
      valueMin: seriesProps.valueMin,
      valueMax: seriesProps.valueMax
    };
  }
  processImpl(datum, seriesProps, xAxis, yAxis, datapoint, context) {
    const { seriesMap, seriesToMetadata } = context;
    const seriesId = this.getSeriesId(datum, seriesProps);
    let currentSeries = seriesMap.get(seriesId);
    if (!currentSeries) {
      currentSeries = this.createSeries(seriesId, seriesProps, xAxis, yAxis);
      this.registerSeries(currentSeries, seriesId, context);
      const metadata = this.createSeriesMetadata(
        datum,
        seriesProps,
        xAxis,
        yAxis,
        seriesId,
        context
      );
      seriesToMetadata.set(currentSeries, {
        ...metadata,
        isVisible: context.filteredSeriesIds !== void 0 ? false : void 0
      });
    }
    const datapointMetadata = { isGap: true };
    if ((0, import_type_guards.isStringOrValidNumber)(datapoint.value)) {
      const valueDomain = (0, import_get_value_scale_domain.getValueScaleDomain)(
        datapoint.value,
        seriesProps.id,
        this.seriesIdToValueScaleDomain
      );
      this.seriesIdToValueScaleDomain.set(seriesProps.id, valueDomain);
      datapointMetadata.isGap = false;
      const tempSeriesMetadata = (0, import_safe_get.safeGet)(seriesToMetadata, currentSeries);
      tempSeriesMetadata.valueType ??= (0, import_get_legend_type.getLegendType)(datapoint.value);
      const isDatapointVisible = tempSeriesMetadata.valueType === import_xy_chart_internals.RANGE_LEGEND_TYPE || context.filteredSeriesIds === void 0 || context.filteredSeriesIds.has(`${seriesId}--${datapoint.value}`);
      datapointMetadata.isVisible = isDatapointVisible;
      tempSeriesMetadata.isVisible ||= isDatapointVisible;
    }
    this.registerDatapoint(
      datapoint,
      datapointMetadata,
      currentSeries,
      context
    );
    const datapointIsVisible = datapointMetadata.isVisible === void 0 || datapointMetadata.isVisible === true;
    if (datapointIsVisible) {
      context.axisWithSeries.add(xAxis.id);
      context.axisWithSeries.add(yAxis.id);
      if (!datapointMetadata.isGap) {
        this.addValue({
          series: currentSeries,
          datapoint
        });
      }
    }
    this.updateAxisDomains(
      datapoint,
      xAxis,
      yAxis,
      context.axisIdToScaleDomain,
      datapointIsVisible
    );
  }
  getColor(colorPalette, fallbackColorPalette) {
    const validColorPalette = (0, import_get_valid_color_palette.getValidColorPalette)(
      colorPalette,
      fallbackColorPalette ?? import_defaults.DEFAULT_COLOR_PALETTE_RECT
    );
    return {
      colorPalette: validColorPalette,
      colorIsFromSlot: validColorPalette === colorPalette
    };
  }
  getDatapoint(datum, seriesProps, xAxis, yAxis) {
    return (0, import_get_datapoints.getRectDatapoint)(datum, seriesProps, xAxis, yAxis);
  }
}
