// @dynatrace/strato-components/charts/xy-chart/processing-strategies/strategies/stackable-series.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var stackable_series_exports = {};
__export(stackable_series_exports, {
  StackableSeriesHelper: () => StackableSeriesHelper
});
module.exports = __toCommonJS(stackable_series_exports);
var import_stack_datapoints = require("../utils/stacking/stack-datapoints.js");
class StackableSeriesHelper {
  static STARTING_POINT = 0;
  stackGroups;
  stackDomainData;
  axisStats;
  constructor() {
    this.stackGroups = /* @__PURE__ */ new Map();
    this.stackDomainData = /* @__PURE__ */ new Map();
    this.axisStats = /* @__PURE__ */ new Map();
  }
  /**
   * Accumulates total positive and negative values per stack key.
   * Should be called during the preprocess phase before processing individual datapoints.
   * This is needed so that binDomain can be included in metadata for relative stacking.
   *
   * Also records which series contributed to each axisKey so that `getStackBinDomain` can
   * fall back to a cross-bin reference for single-series axes. Single-series stacks would
   * otherwise collapse to 100% in every bin.
   */
  processStackDomain(datapoint, chartOrientation, xAxis, yAxis, seriesId) {
    const { stackKey, value } = (0, import_stack_datapoints.getStackingData)(
      chartOrientation,
      datapoint,
      xAxis,
      yAxis
    );
    const axisKey = (0, import_stack_datapoints.getCombinedAxesOrientationKey)(
      xAxis.id,
      yAxis.id,
      chartOrientation
    );
    const stats = this.getOrInitAxisStats(axisKey);
    stats.seriesIds.add(seriesId);
    if (typeof value === "number") {
      const acc = this.stackDomainData.get(stackKey) ?? {
        totalPositive: StackableSeriesHelper.STARTING_POINT,
        totalNegative: StackableSeriesHelper.STARTING_POINT,
        axisKey
      };
      if (value >= 0) {
        acc.totalPositive += value;
      } else {
        acc.totalNegative += Math.abs(value);
      }
      this.stackDomainData.set(stackKey, acc);
      const binAbsValue = Math.max(acc.totalPositive, acc.totalNegative);
      if (binAbsValue > stats.maxBinAbsValue) {
        stats.maxBinAbsValue = binAbsValue;
      }
    }
  }
  /**
   * Returns the total bin domain for a given stack key.
   *
   * For axes with more than one contributing series the result is the per-bin total height
   * of the stack (|positive| + |negative| within the bin). For axes with only one contributing
   * series the result collapses to a cross-bin reference: the largest per-bin absolute total
   * across all bins of that axis. This keeps single-series stacks consistent with the regular
   * relative behavior, where the tallest bar represents 100% and the rest scale proportionally.
   */
  getStackBinDomain(stackKey) {
    const domainData = this.stackDomainData.get(stackKey);
    if (!domainData) {
      return void 0;
    }
    const stats = this.axisStats.get(domainData.axisKey);
    if (stats && stats.seriesIds.size === 1) {
      return stats.maxBinAbsValue;
    }
    return domainData.totalPositive + domainData.totalNegative;
  }
  getOrInitAxisStats(axisKey) {
    let stats = this.axisStats.get(axisKey);
    if (!stats) {
      stats = { seriesIds: /* @__PURE__ */ new Set(), maxBinAbsValue: 0 };
      this.axisStats.set(axisKey, stats);
    }
    return stats;
  }
  /**
   * Processes a datapoint for stacking, calculating the offset and updating stack groups.
   * Returns the calculated datapoint with stacking offsets applied and enriched metadata.
   */
  processStackedDatapoint(datapoint, chartOrientation, xAxis, yAxis, datapointMetadata) {
    const { stackKey, value } = (0, import_stack_datapoints.getStackingData)(
      chartOrientation,
      datapoint,
      xAxis,
      yAxis
    );
    if (typeof value !== "number") {
      return {
        calcDatapoint: {
          x0: datapoint.x0,
          x1: datapoint.x1,
          y0: datapoint.y0,
          y1: datapoint.y1
        },
        datapointMetadata,
        shouldContinue: true
      };
    }
    const prevValue = this.stackGroups.get(stackKey) ?? {
      minBoundary: StackableSeriesHelper.STARTING_POINT,
      maxBoundary: StackableSeriesHelper.STARTING_POINT
    };
    const binDomain = this.getStackBinDomain(stackKey);
    const updatedMetadata = {
      ...datapointMetadata,
      accumulator: value > 0 ? prevValue.maxBoundary : prevValue.minBoundary,
      binDomain
    };
    const stackingOffset = value > 0 ? prevValue.maxBoundary : prevValue.minBoundary;
    const calcDatapoint = {
      x0: chartOrientation === "horizontal" && typeof datapoint.x0 === "number" ? datapoint.x0 + stackingOffset : datapoint.x0,
      x1: chartOrientation === "horizontal" && typeof datapoint.x1 === "number" ? datapoint.x1 + stackingOffset : datapoint.x1,
      y0: chartOrientation === "vertical" && typeof datapoint.y0 === "number" ? datapoint.y0 + stackingOffset : datapoint.y0,
      y1: chartOrientation === "vertical" && typeof datapoint.y1 === "number" ? datapoint.y1 + stackingOffset : datapoint.y1
    };
    this.stackGroups.set(stackKey, {
      ...prevValue,
      ...(0, import_stack_datapoints.getStackedValue)(value, prevValue)
    });
    return {
      calcDatapoint,
      datapointMetadata: updatedMetadata,
      shouldContinue: false
    };
  }
}
