// @dynatrace/strato-components/esm/charts/pie/components/PieChartToolbar/components/DownloadDataButton.js
import { jsx } from "react/jsx-runtime";
import { useRef } from "react";
import { useIntl } from "react-intl";
import { DownloadIcon } from "@dynatrace/strato-icons";
import { ChartToolbarRenderer } from "../../../../core/components/toolbar/ChartToolbarRenderer.js";
import { useA11yToolbarTabIndex } from "../../../../core/hooks/useA11yToolbarTabIndex.js";
import { usePieChartRawData } from "../../../context/PieChartRawDataContext.js";
import { PIE_CHART_MESSAGES } from "../../../messages.js";
import { buildAndDownloadCsv } from "../../../utils/build-and-download-csv.js";
const DownloadDataButton = () => {
  const downloadDataButtonRef = useRef(null);
  const tabIndex = useA11yToolbarTabIndex("download");
  const intl = useIntl();
  const rawData = usePieChartRawData();
  const handleDownloadDataKeyboard = (event) => {
    if (event.key === "Enter") {
      buildAndDownloadCsv(rawData);
    }
  };
  const handleDownloadDataToolbar = () => {
    buildAndDownloadCsv(rawData);
  };
  return /* @__PURE__ */ jsx(
    ChartToolbarRenderer.Control,
    {
      ref: downloadDataButtonRef,
      prefixIcon: /* @__PURE__ */ jsx(DownloadIcon, {}),
      tooltipText: intl.formatMessage(PIE_CHART_MESSAGES.downloadAsCsv),
      testId: "chart-toolbar-download",
      onKeyDown: handleDownloadDataKeyboard,
      onClick: handleDownloadDataToolbar,
      tabIndex
    }
  );
};
DownloadDataButton["displayName"] = "DownloadDataButton";
export {
  DownloadDataButton
};
