// @dynatrace/strato-components/esm/charts/tree-map/store/selectors.js
import { useContext } from "react";
import { TreeMapStateContext } from "./store.js";
import { useStoreSelector } from "../../../core/store/selector.js";
const useSelector = (selector) => useStoreSelector(TreeMapStateContext, selector);
const useSetState = () => {
  const { setState } = useContext(TreeMapStateContext);
  return setState;
};
const useTreeMapState = () => useSelector((state) => state);
const useHighlightedState = () => useSelector((state) => state.highlightedItem);
const useGroupsSelectedState = () => useSelector((state) => state.groupsSelected);
const useLeafNodesState = () => useSelector((state) => state.hierarchyLeafNodes);
const useClusterNodesState = () => useSelector((state) => state.hierarchyClusterNodes);
const useTreeMapToolbarVisibleState = () => useSelector((state) => state.overlay.toolbarVisible);
const useColorMap = () => useSelector((state) => state.colorMap);
export {
  useClusterNodesState,
  useColorMap,
  useGroupsSelectedState,
  useHighlightedState,
  useLeafNodesState,
  useSelector,
  useSetState,
  useTreeMapState,
  useTreeMapToolbarVisibleState
};
