// @dynatrace/strato-components/content/release-phase-indicator/messages.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var messages_exports = {};
__export(messages_exports, {
  releasePhaseMessages: () => releasePhaseMessages
});
module.exports = __toCommonJS(messages_exports);
var import_react_intl = require("react-intl");
const releasePhaseMessages = (0, import_react_intl.defineMessages)({
  preview: {
    defaultMessage: "Preview",
    description: "Chip label indicating a feature or an app is in preview release phase",
    id: "92og/kC9Mmwxmmg0"
  },
  new: {
    defaultMessage: "New",
    description: "Chip label indicating a feature is newly released",
    id: "+Jc1Pz2t25AcHil0"
  },
  deprecated: {
    defaultMessage: "Deprecated",
    description: "Chip label indicating a feature is deprecated",
    id: "VwnLYpww5VUT3cdu"
  }
});
