// @dynatrace/strato-components/content/release-phase-indicator/ReleasePhaseIndicator.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ReleasePhaseIndicator_exports = {};
__export(ReleasePhaseIndicator_exports, {
  ReleasePhaseIndicator: () => ReleasePhaseIndicator
});
module.exports = __toCommonJS(ReleasePhaseIndicator_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_messages = require("./messages.js");
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useSafeForwardProps = require("../../core/hooks/useSafeForwardProps.js");
var import_Chip = require("../chip/Chip.js");
const COMPONENT_NAME = "ReleasePhaseIndicator";
const phaseColorMap = {
  preview: "primary",
  new: "primary",
  deprecated: "warning"
};
const phaseLabelMap = {
  preview: import_messages.releasePhaseMessages.preview,
  new: import_messages.releasePhaseMessages.new,
  deprecated: import_messages.releasePhaseMessages.deprecated
};
const ReleasePhaseIndicator = (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    phase,
    id,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    ...remainingProps
  } = props;
  const intl = (0, import_react_intl.useIntl)();
  const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const ariaLabelingProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(restProps);
  const forwardProps = (0, import_useSafeForwardProps.useSafeForwardProps)(restProps, ariaLabelingProps);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Chip.Chip,
    {
      id,
      ref: forwardedRef,
      size: "condensed",
      variant: "emphasized",
      color: phaseColorMap[phase],
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...ariaLabelingProps,
      ...behavioralTrackingProps,
      ...forwardProps,
      "data-dt-component": COMPONENT_NAME,
      children: intl.formatMessage(phaseLabelMap[phase])
    }
  );
});
ReleasePhaseIndicator.displayName = "ReleasePhaseIndicator";
