// @dynatrace/strato-components/buttons/notify-button/NotifyButton.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var NotifyButton_exports = {};
__export(NotifyButton_exports, {
  NotifyButton: () => NotifyButton
});
module.exports = __toCommonJS(NotifyButton_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_is = require("react-is");
var import_Content = require("./components/Content.js");
var import_Trigger = require("./components/Trigger.js");
var import_NotifyButtonContext = require("./contexts/NotifyButtonContext.js");
var styles = __toESM(require("./NotifyButton.sty.js"));
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useId = require("../../core/hooks/useId.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_is_node_environment = require("../../core/utils/is-node-environment.js");
var import_merge_props = require("../../core/utils/merge-props.js");
var import_Flex = require("../../layouts/flex/Flex.js");
var import_Content2 = require("../../notifications/notification-settings/components/Content.js");
var import_PermissionInfo = require("../../notifications/notification-settings/components/PermissionInfo.js");
var import_ToggleButton = require("../../notifications/notification-settings/components/ToggleButton.js");
var import_NotificationSettings = require("../../notifications/notification-settings/NotificationSettings.js");
var import_useOverlayWithTrigger = require("../../overlays/hooks/useOverlayWithTrigger.js");
var import_Overlay = require("../../overlays/overlay/Overlay.js");
const COMPONENT_NAME = "NotifyButton";
function getContentSlot(children) {
  let content;
  import_react.Children.forEach(children, (child) => {
    if ((0, import_react_is.isElement)(child)) {
      switch (child.type) {
        case import_Content.Content:
          content = child;
          break;
        default: {
          if ((0, import_is_node_environment.isDev)()) {
            console.warn("Only Content is a valid child of the NotifyButton.");
          }
          break;
        }
      }
    } else if (child) {
      if ((0, import_is_node_environment.isDev)()) {
        console.warn("The used child is not supported by the NotifyButton.");
      }
    }
  });
  return content;
}
const RootNotifyButton = (0, import_react.forwardRef)(
  (props, ref) => {
    const {
      children,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      className: consumerClassName,
      style: consumerStyle,
      disabled,
      size = "default",
      variant = "default",
      type,
      config,
      readOnly,
      showLabel = false,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
    const overlayAriaLabelledBy = (0, import_useId.useId)();
    const {
      targetRef,
      triggerRef,
      overlayRef,
      overlayProps,
      overlayState,
      overlayTriggerProps
    } = (0, import_useOverlayWithTrigger.useOverlayWithTrigger)({
      placement: "bottom-start",
      placementFallbacks: [
        "bottom-start",
        "bottom-end",
        "top-start",
        "top-end"
      ],
      offset: 4,
      disabled: disabled || readOnly,
      "aria-labelledby": overlayAriaLabelledBy
    });
    const rootRef = (0, import_useMergeRefs.useMergeRefs)([ref, triggerRef, targetRef]);
    const contentSlot = (0, import_react.useMemo)(() => getContentSlot(children), [children]);
    const notifyButtonContextValue = (0, import_react.useMemo)(
      () => ({
        closeOverlay: overlayState.close,
        overlayAriaLabelledBy
      }),
      [overlayAriaLabelledBy, overlayState.close]
    );
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_NotifyButtonContext.NotifyButtonContext.Provider, { value: notifyButtonContextValue, children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_NotificationSettings.NotificationProvider, { type, config, children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_Trigger.NotifyButtonTrigger,
        {
          ref: rootRef,
          "data-testid": dataTestId,
          "data-dtrum-mask": dataDtrumMask,
          "data-dtrum-allow": dataDtrumAllow,
          className: consumerClassName,
          style: consumerStyle,
          disabled,
          size,
          variant,
          showLabel,
          triggerProps: overlayTriggerProps,
          "aria-disabled": readOnly,
          ...(0, import_merge_props.mergeProps)(restProps, behavioralTrackingProps),
          "data-dt-component": COMPONENT_NAME
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_Overlay.Overlay,
        {
          ref: overlayRef,
          isOpen: overlayState.isOpen,
          className: styles.overlay,
          overlayProps,
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { flexDirection: "column", gap: 12, children: [
            contentSlot ?? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Content2.Content, {}),
            /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { alignItems: "center", children: [
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ToggleButton.NotificationToggleButton, {}),
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_PermissionInfo.NotificationPermissionInfo, {})
            ] })
          ] })
        }
      )
    ] }) });
  }
);
RootNotifyButton.displayName = COMPONENT_NAME;
const NotifyButton = Object.assign(RootNotifyButton, {
  Content: import_Content.Content
});
