// @dynatrace/strato-components/buttons/notify-button/notify-button-messages.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var notify_button_messages_exports = {};
__export(notify_button_messages_exports, {
  notifyButtonTranslations: () => notifyButtonTranslations
});
module.exports = __toCommonJS(notify_button_messages_exports);
var import_react_intl = require("react-intl");
const notifyButtonTranslations = (0, import_react_intl.defineMessages)({
  notifications: {
    defaultMessage: "Notifications",
    id: "GH0lEEpNg7HJ9sA6",
    description: "Label text shown in the notify button trigger"
  },
  toggleOverlay: {
    defaultMessage: "Toggle notify button notifications overlay",
    id: "jishWwgadkdK/YkI",
    description: "Aria label used for the notify button trigger"
  }
});
