// @dynatrace/strato-components/esm/editors/code-editor/CodeEditor.js
import { jsx } from "react/jsx-runtime";
import { forwardRef, useMemo } from "react";
import { useIntl } from "react-intl";
import { PLACEHOLDER_MESSAGES } from "./codeEditorMessages.js";
import { codeEditorLanguages } from "./languages/languages.js";
import { useAriaLabelingProps } from "../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useSafeForwardProps } from "../../core/hooks/useSafeForwardProps.js";
import { BaseCodeEditor } from "../base-code-editor/BaseCodeEditor.js";
const COMPONENT_NAME = "CodeEditor";
const CodeEditor = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const intl = useIntl();
  const {
    id,
    controlState,
    defaultFolding,
    folding,
    onFoldingChange,
    spellCheck = false,
    language = "other",
    placeholder = intl.formatMessage(PLACEHOLDER_MESSAGES[language]),
    readOnly = false,
    lineWrap = false,
    fullHeight = false,
    value = "",
    onChange,
    onBlur,
    onFocus,
    gutterConfiguration,
    size = "default",
    required = false,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
  const ariaLabelingProps = useAriaLabelingProps(restProps);
  const forwardProps = useSafeForwardProps(restProps, ariaLabelingProps);
  const config = useMemo(() => {
    return getLanguageConfig(codeEditorLanguages[language], readOnly);
  }, [language, readOnly]);
  return /* @__PURE__ */ jsx(
    BaseCodeEditor,
    {
      id,
      ref: forwardedRef,
      placeholder,
      controlState,
      readOnly,
      folding,
      defaultFolding,
      onFoldingChange,
      lineWrap,
      value,
      onChange,
      spellCheck,
      fullHeight,
      configuration: config,
      onBlur,
      onFocus,
      gutterConfiguration,
      size,
      required,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...ariaLabelingProps,
      ...behavioralTrackingProps,
      ...forwardProps,
      "data-dt-component": COMPONENT_NAME
    }
  );
});
function getLanguageConfig(languageConfig, readOnly) {
  return {
    language: languageConfig.language,
    autocompleteOptions: readOnly ? void 0 : {
      activateOnTyping: true,
      providers: languageConfig.autocompleteOptions?.providers ?? []
    },
    extendWith: languageConfig.extendWith
  };
}
CodeEditor.displayName = COMPONENT_NAME;
export {
  CodeEditor,
  getLanguageConfig
};
