// @dynatrace/strato-components/esm/forms/_base-input/BaseInputContent.js
import { jsx } from "react/jsx-runtime";
import { forwardRef } from "react";
import * as styles from "./BaseInputContent.sty.js";
import { TextEllipsisAroundChildren } from "../../core/components/TextEllipsisAroundChildren.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { Flex } from "../../layouts/flex/Flex.js";
const BaseInputContent = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const { children, ...rest } = props;
  const [behavioralTrackingProps] = useBehavioralTrackingProps(rest);
  return /* @__PURE__ */ jsx(
    Flex,
    {
      ref: forwardedRef,
      flexGrow: 0,
      alignItems: "center",
      alignSelf: "stretch",
      gap: 8,
      paddingLeft: 8,
      paddingRight: 8,
      className: styles.content,
      ...behavioralTrackingProps,
      children: /* @__PURE__ */ jsx(TextEllipsisAroundChildren, { children })
    }
  );
});
const BaseInputPrefix = /* @__PURE__ */ forwardRef((props, forwardedRef) => /* @__PURE__ */ jsx(BaseInputContent, { ref: forwardedRef, ...props }));
const BaseInputSuffix = /* @__PURE__ */ forwardRef((props, forwardedRef) => /* @__PURE__ */ jsx(BaseInputContent, { ref: forwardedRef, ...props }));
const BaseInputReservedSuffix = /* @__PURE__ */ forwardRef((props, forwardedRef) => /* @__PURE__ */ jsx(BaseInputContent, { ref: forwardedRef, ...props }));
BaseInputPrefix.displayName = "Input.Prefix";
BaseInputSuffix.displayName = "Input.Suffix";
export {
  BaseInputPrefix,
  BaseInputReservedSuffix,
  BaseInputSuffix
};
