// @dynatrace/strato-components/esm/forms/_base-input/BaseInput.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import {
  Children,
  forwardRef,
  isValidElement,
  useImperativeHandle,
  useMemo,
  useRef
} from "react";
import { useIntl } from "react-intl";
import * as styles from "./BaseInput.sty.js";
import { BaseInputButton } from "./BaseInputButton.js";
import {
  BaseInputPrefix,
  BaseInputReservedSuffix,
  BaseInputSuffix
} from "./BaseInputContent.js";
import { BaseInputContext } from "./contexts/BaseInputContext.js";
import { useFormControlState } from "../../core/hooks/useFormControlState.js";
import { useSplitProps } from "../../core/hooks/useSplitProps.js";
import { useFocusRing } from "../../core/styles/useFocusRing.js";
import { isSlot } from "../../core/utils/is-slot.js";
import { mergeProps as _mergeProps } from "../../core/utils/merge-props.js";
import { Divider } from "../../layouts/divider/Divider.js";
import { Flex } from "../../layouts/flex/Flex.js";
import { inputGroupClassName } from "../../layouts/input-group/InputGroup.sty.js";
import { InputGroup } from "../../layouts/input-group/InputGroup.js";
import { InternalBaseItem as BaseItem } from "../forms-core/components/form-field-messages/FormFieldMessageBaseItem.js";
import { useFormFieldMessagesContext } from "../forms-core/hooks/useFormFieldMessagesContext.js";
import { useFormFieldValidation } from "../forms-core/hooks/useFormFieldValidation.js";
import { useSyncedFormControlProps } from "../forms-core/hooks/useSyncedFormControlProps.js";
import { FormFieldMessagesAction } from "../forms-core/state/formfield-messages-state-reducer.js";
import { useAutofillEvent } from "../hooks/useAutofillEvent.js";
import { useInputHint } from "../hooks/useInputHint.js";
import { getPlaceholderMessage } from "./baseInputMessages.js";
import { ClearButton } from "./ClearButton.js";
function getSlots(children) {
  const validSlots = Children.toArray(children).filter(
    isValidElement
  );
  let prefix = { buttons: [], rest: [], props: {} };
  let suffix = { buttons: [], rest: [], props: {} };
  let reservedSuffix = { buttons: [], rest: [], props: {} };
  for (const slot of validSlots) {
    const buttons = [];
    const rest = [];
    const slotChildren = Children.toArray(slot.props.children);
    for (const child of slotChildren) {
      if (isSlot(child, BaseInputButton)) {
        buttons.push(child);
      } else {
        rest.push(child);
      }
    }
    const { children: _children, ...slotProps } = slot.props;
    if (slot.type === BaseInputPrefix) {
      prefix = { buttons, rest, props: slotProps };
    }
    if (slot.type === BaseInputSuffix) {
      suffix = { buttons, rest, props: slotProps };
    }
    if (slot.type === BaseInputReservedSuffix) {
      reservedSuffix = { buttons, rest, props: slotProps };
    }
  }
  return { prefix, suffix, reservedSuffix };
}
const BaseInputRoot = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const intl = useIntl();
  const defaultPlaceholder = intl.formatMessage(
    getPlaceholderMessage(props.type)
  );
  const {
    children,
    id: propId,
    disabled: propDisabled,
    required: propRequired,
    "aria-disabled": ariaDisabledProp,
    variant = "default",
    controlState,
    readOnly: propReadOnly,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "aria-label": ariaLabelProp,
    "aria-labelledby": ariaLabelledByProp,
    "aria-describedby": ariaDescribedByProp,
    "aria-details": ariaDetailsProp,
    className: consumerClassName,
    style: consumerStyle,
    // eslint-disable-next-line @typescript-eslint/no-deprecated
    onClear,
    type,
    placeholder = defaultPlaceholder,
    value,
    pattern,
    max,
    min,
    step,
    maxLength,
    minLength,
    ...remainingProps
  } = props;
  const inputValue = value ?? "";
  const { disabled, required, id, ariaDisabled, readOnly } = useSyncedFormControlProps({
    propDisabled,
    propRequired,
    propId,
    propAriaDisabled: ariaDisabledProp,
    propReadOnly
  });
  const disabledOrAriaDisabled = disabled || ariaDisabled || false;
  const [hasFormControlError, hint] = useFormControlState(controlState);
  const { inputProps, hintProps } = useInputHint({
    hint,
    hasError: hasFormControlError
  });
  const { state, dispatch } = useFormFieldMessagesContext();
  const hasError = hasFormControlError || state.showError;
  const { isAutoFilled, onAnimationStart } = useAutofillEvent();
  const inputRef = useRef(null);
  const wrapperRef = useRef(null);
  const { focusProps, focusClassName } = useFocusRing({
    color: hasError && !(disabled || ariaDisabled) ? "critical" : "neutral",
    ignoreModality: true,
    disabled
  });
  const { onChange, onBlur, onInvalid, ariaProps } = useFormFieldValidation(
    inputRef,
    {
      required,
      pattern,
      max,
      min,
      step,
      maxLength,
      minLength,
      type,
      value: inputValue
    },
    {
      "aria-label": ariaLabelProp,
      "aria-labelledby": ariaLabelledByProp,
      "aria-describedby": ariaDescribedByProp,
      "aria-details": ariaDetailsProp
    }
  );
  const eventListener = { onChange, onBlur, onInvalid };
  useImperativeHandle(
    forwardedRef,
    () => ({
      element: wrapperRef.current,
      inputRef: inputRef.current,
      focus() {
        inputRef.current?.focus();
      },
      validate() {
        inputRef.current?.checkValidity();
        dispatch?.({
          type: FormFieldMessagesAction.SETISTOUCHED,
          payload: { isTouched: true }
        });
      }
    }),
    [dispatch]
  );
  const context = useMemo(
    () => ({ variant, hasError, disabled, ariaDisabled }),
    [variant, hasError, disabled, ariaDisabled]
  );
  const slots = useMemo(() => getSlots(children), [children]);
  const hasInnerPrefix = slots.prefix.rest.length > 0;
  const hasInnerSuffix = slots.suffix.rest.length > 0;
  const hasInnerReservedSuffix = slots.reservedSuffix.rest.length > 0;
  const hasOnClear = onClear !== void 0;
  const showClearButton = hasOnClear && inputValue !== "" && !hasInnerReservedSuffix;
  const {
    matchedProps: dataAttributesProps,
    remainingProps: remainingInputProps
  } = useSplitProps(remainingProps, [(key) => key.startsWith("data-")]);
  return /* @__PURE__ */ jsxs(
    Flex,
    {
      gap: 4,
      flexDirection: "column",
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ref: wrapperRef,
      className: consumerClassName,
      style: consumerStyle,
      ...dataAttributesProps,
      children: [
        /* @__PURE__ */ jsx(
          InputGroup,
          {
            style: {
              gridAutoColumns: `${"auto ".repeat(
                slots.prefix.buttons.length
              )}1fr${" auto".repeat(slots.suffix.buttons.length)}`
            },
            className: clsx(styles.wrapper, styles.variant[variant], {
              [styles.hasError]: hasError,
              [styles.disabled]: disabledOrAriaDisabled,
              [styles.readOnly]: readOnly
            }),
            children: /* @__PURE__ */ jsxs(BaseInputContext.Provider, { value: context, children: [
              slots.prefix.buttons,
              /* @__PURE__ */ jsxs(
                "div",
                {
                  className: clsx(
                    styles.container,
                    inputGroupClassName,
                    focusClassName,
                    {
                      [styles.isAutoFilled]: isAutoFilled,
                      [`${styles.containerPadding.none}`]: hasInnerReservedSuffix && hasInnerPrefix,
                      [`${styles.containerPadding.rightNone}`]: hasInnerReservedSuffix && !hasInnerPrefix,
                      [`${styles.containerPadding.leftUnset}`]: hasInnerPrefix,
                      [`${styles.containerPadding.rightUnset}`]: hasInnerSuffix
                    }
                  ),
                  onClick: () => inputRef.current?.focus(),
                  children: [
                    hasInnerPrefix && /* @__PURE__ */ jsx(BaseInputPrefix, { ...slots.prefix.props, children: slots.prefix.rest }),
                    /* @__PURE__ */ jsx(
                      "input",
                      {
                        id,
                        ref: inputRef,
                        required,
                        disabled: !!disabled,
                        "aria-disabled": ariaDisabled,
                        onAnimationStart,
                        readOnly,
                        className: styles.input,
                        placeholder,
                        type,
                        value: inputValue,
                        pattern,
                        max,
                        min,
                        step,
                        maxLength,
                        minLength,
                        ..._mergeProps(
                          remainingInputProps,
                          focusProps,
                          controlState ? { "aria-label": ariaLabelProp, ...inputProps } : ariaProps,
                          !controlState ? eventListener : {}
                          // escape hatch to not interfere with the old implementation
                        )
                      }
                    ),
                    hasInnerSuffix && /* @__PURE__ */ jsx(BaseInputSuffix, { ...slots.suffix.props, children: slots.suffix.rest }),
                    !readOnly && /* @__PURE__ */ jsx(Fragment, { children: hasInnerReservedSuffix ? /* @__PURE__ */ jsxs(Fragment, { children: [
                      hasInnerSuffix && /* @__PURE__ */ jsx(
                        Divider,
                        {
                          orientation: "vertical",
                          className: styles.divider
                        }
                      ),
                      slots.reservedSuffix.rest
                    ] }) : /* @__PURE__ */ jsxs(Fragment, { children: [
                      hasInnerSuffix && showClearButton && /* @__PURE__ */ jsx(
                        Divider,
                        {
                          orientation: "vertical",
                          className: styles.divider
                        }
                      ),
                      showClearButton && /* @__PURE__ */ jsx(ClearButton, { onClear })
                    ] }) })
                  ]
                }
              ),
              slots.suffix.buttons
            ] })
          }
        ),
        controlState && hint && /* @__PURE__ */ jsx(
          BaseItem,
          {
            variant: hasError ? "error" : "hint",
            disabled: disabledOrAriaDisabled,
            ...hintProps,
            children: hint
          }
        )
      ]
    }
  );
});
const BaseInput = Object.assign(BaseInputRoot, {
  Prefix: BaseInputPrefix,
  Suffix: BaseInputSuffix,
  ReservedSuffix: BaseInputReservedSuffix,
  Button: BaseInputButton
});
BaseInput.displayName = "BaseInput";
export {
  BaseInput
};
