// @dynatrace/strato-components/esm/forms/_base-input/BaseInputButton.js
import { jsx } from "react/jsx-runtime";
import {
  forwardRef,
  useMemo
} from "react";
import { useBaseInputContext } from "./hooks/useBaseInputContext.js";
import { Button } from "../../buttons/button/Button.js";
import { useAriaLabelingProps as _useAriaLabelingProps } from "../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useId } from "../../core/hooks/useId.js";
import { useSafeForwardProps as _useSafeForwardProps } from "../../core/hooks/useSafeForwardProps.js";
import { mergeProps } from "../../core/utils/merge-props.js";
import { Tooltip } from "../../overlays/tooltip/Tooltip.js";
const BaseInputButton = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const {
    children,
    tooltipText,
    tooltipBehavioralTrackingProps: tooltipTrackingProps,
    disabled: propDisabled,
    id,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "aria-disabled": ariaDisabledProp,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restPropsWithoutTracking] = useBehavioralTrackingProps(remainingProps);
  const {
    hasError = false,
    disabled,
    variant,
    ariaDisabled: contextAriaDisabled
  } = useBaseInputContext();
  const ariaDisabled = ariaDisabledProp || contextAriaDisabled;
  const tooltipGeneratedId = useId();
  const tooltipId = tooltipText ? id ?? tooltipGeneratedId : void 0;
  const { "aria-labelledby": ariaLabelledByProp, ...remainingAriaProps } = _useAriaLabelingProps(restPropsWithoutTracking);
  const forwardProps = _useSafeForwardProps(restPropsWithoutTracking, {
    ariaLabelledByProp,
    ...remainingAriaProps
  });
  const ariaLabelledBy = useMemo(
    () => tooltipId || ariaLabelledByProp ? `${tooltipId ? tooltipId + " " : ""}${ariaLabelledByProp ?? ""}` : void 0,
    [ariaLabelledByProp, tooltipId]
  );
  const button = useMemo(
    () => /* @__PURE__ */ jsx(
      Button,
      {
        ref: forwardedRef,
        ...forwardProps,
        variant: variant === "minimal" ? "default" : "emphasized",
        disabled: propDisabled ?? disabled,
        "aria-disabled": ariaDisabled,
        color: hasError && !disabled ? "critical" : "neutral",
        "aria-labelledby": ariaLabelledBy,
        className: consumerClassName,
        style: consumerStyle,
        "data-testid": dataTestId,
        ...mergeProps(remainingAriaProps, behavioralTrackingProps),
        children: /* @__PURE__ */ jsx(Button.Prefix, { children })
      }
    ),
    [
      ariaLabelledBy,
      behavioralTrackingProps,
      children,
      consumerClassName,
      consumerStyle,
      dataTestId,
      disabled,
      forwardProps,
      forwardedRef,
      hasError,
      propDisabled,
      remainingAriaProps,
      variant,
      ariaDisabled
    ]
  );
  return tooltipText ? /* @__PURE__ */ jsx(
    Tooltip,
    {
      id: tooltipId,
      text: tooltipText,
      placement: "top-end",
      ...tooltipTrackingProps,
      children: button
    }
  ) : button;
});
BaseInputButton.displayName = "BaseInput.Button";
export {
  BaseInputButton
};
