// @dynatrace/strato-components/esm/core/time/timeframeStore/actions/calendar.js
import { isBefore } from "date-fns";
import { CALENDAR_ERROR_MESSAGES } from "../../../../forms/calendar/calendarMessages.js";
import { getTimeValueFromSimpleValue } from "../../../../forms/forms-core/time/helpers.js";
import { isFullTimeframe, isTimeframe } from "../../helpers.js";
import { validate } from "../../validation.js";
function getTimeframeFromCalendarSelection(inputValue, state) {
  if (!isTimeframe(state.calendarValue, "range")) {
    return null;
  }
  if (inputValue === null) {
    return null;
  }
  const stringDate = inputValue.toISOString();
  const calendarRangeValue = state.calendarValue;
  const { from, to } = calendarRangeValue ?? {};
  let position = "from";
  let newTimeframe = {
    from: null,
    to: null
  };
  if (from && !to) {
    const isToBeforeFrom = isBefore(stringDate, from.absoluteDate);
    if (!isToBeforeFrom) {
      newTimeframe = {
        from,
        to: null
      };
      position = "to";
    }
    if (isToBeforeFrom) {
      newTimeframe = {
        from: null,
        to: getTimeValueFromSimpleValue(
          from.absoluteDate,
          state.timeOffset,
          "to",
          void 0,
          state.min,
          state.max,
          state.precision
        )
      };
    }
  }
  const newTimeValue = getTimeValueFromSimpleValue(
    stringDate,
    state.timeOffset,
    position,
    void 0,
    state.min,
    state.max,
    state.precision
  );
  newTimeframe[position] = newTimeValue;
  return newTimeframe;
}
function getCalendarActions(isControlled, type, formFieldMessagesDispatch, onChangeProp) {
  return (set) => ({
    setCalendarValue: (inputValue) => {
      set((state) => {
        if (type === "range") {
          const newTimeframe = getTimeframeFromCalendarSelection(
            inputValue,
            state
          );
          if (isFullTimeframe(newTimeframe) || newTimeframe === null) {
            onChangeProp?.(newTimeframe);
            if (isControlled) {
              return state;
            }
          }
          const calendarError2 = validate(newTimeframe, state.min, state.max, {
            formMessageDispatch: formFieldMessagesDispatch,
            formerError: state.calendarError,
            precision: state.precision,
            showError: false,
            checkOrder: true,
            ref: state.calendarRef,
            valueMode: "timeframeSelector",
            errorMessages: CALENDAR_ERROR_MESSAGES
          });
          return {
            calendarValue: newTimeframe,
            calendarError: calendarError2
          };
        }
        const newTimeValue = getTimeValueFromSimpleValue(
          inputValue,
          state.timeOffset,
          void 0,
          void 0,
          state.min,
          state.max,
          state.precision
        ) ?? null;
        onChangeProp?.(
          newTimeValue?.value ?? null
        );
        if (isControlled) {
          return state;
        }
        const calendarError = validate(newTimeValue, state.min, state.max, {
          formMessageDispatch: formFieldMessagesDispatch,
          formerError: state.calendarError,
          precision: state.precision,
          showError: false,
          checkOrder: true,
          ref: state.calendarRef,
          valueMode: "single",
          errorMessages: CALENDAR_ERROR_MESSAGES
        });
        return { calendarValue: newTimeValue, calendarError };
      });
    }
  });
}
export {
  getCalendarActions,
  getTimeframeFromCalendarSelection
};
