// @dynatrace/strato-components/esm/core/time/timeframeStore/actions/dateTimeRangePicker.js
import { getTimeframeFromCalendarSelection } from "./calendar.js";
import { focusElementWithRetry } from "../../../../forms/date-time-picker/utils/focusDateTimeInput.js";
import { removePrecisionFromTimeFrame } from "../../../../forms/forms-core/time/timeParser.js";
import {
  isFullTimeframe,
  isTimeValue,
  mergeRangeValue
} from "../../helpers.js";
import { validate } from "../../validation.js";
function getDateTimeRangePickerActions(formFieldMessagesDispatch, onChangeProp) {
  const actions = (set) => {
    return {
      setCalendarValue: (inputValue) => {
        set((state) => {
          const calendarTimeframe = getTimeframeFromCalendarSelection(
            inputValue,
            state
          );
          if (isFullTimeframe(calendarTimeframe)) {
            state.dateTimeRangePickerRef?.current?.close();
            focusElementWithRetry(
              () => state.dateTimeFromRef.current?.spinButtonHandlers.get("hour")?.element.current,
              state.dateTimeFromRef
            );
            onChangeProp?.(calendarTimeframe);
            return {
              calendarValue: calendarTimeframe,
              dateTimeRangePickerTimeframe: calendarTimeframe
            };
          }
          return { calendarValue: calendarTimeframe };
        });
      },
      setDateTimeRangePickerTimeframe: (value) => {
        return set((state) => {
          if (value === null) {
            onChangeProp?.(null);
            return {
              dateTimeRangePickerTimeframe: null,
              calendarValue: null
            };
          }
          const merged = mergeRangeValue(
            state.dateTimeRangePickerTimeframe,
            value
          );
          const timeframeWithPrecision = removePrecisionFromTimeFrame(merged, state.precision) ?? null;
          if (timeframeWithPrecision?.from === null && timeframeWithPrecision?.to === null) {
            onChangeProp?.(null);
          }
          isFullTimeframe(timeframeWithPrecision) && onChangeProp?.(timeframeWithPrecision);
          const isPartialFrom = isTimeValue(merged.from) && merged.from.absoluteDate === "";
          const isPartialTo = isTimeValue(merged.to) && merged.to.absoluteDate === "";
          if (isPartialFrom || isPartialTo) {
            const errorKey = validate(
              timeframeWithPrecision,
              state.min,
              state.max,
              {
                formMessageDispatch: formFieldMessagesDispatch,
                formerError: state.dateTimeRangePickerError,
                precision: state.precision,
                showError: false,
                checkOrder: true,
                ref: state.dateTimeRangePickerRef,
                expectedFormat: state.formatString,
                valueMode: "dateTimePicker",
                required: state.required
              }
            );
            const calendarValue = {
              from: isPartialFrom ? null : timeframeWithPrecision?.from ?? null,
              to: isPartialTo ? null : timeframeWithPrecision?.to ?? null
            };
            return {
              dateTimeRangePickerTimeframe: timeframeWithPrecision,
              calendarValue,
              calendarError: errorKey,
              dateTimeRangePickerError: errorKey
            };
          }
          const dateTimeRangePickerError = validate(
            timeframeWithPrecision,
            state.min,
            state.max,
            {
              formMessageDispatch: formFieldMessagesDispatch,
              formerError: state.dateTimeRangePickerError,
              precision: state.precision,
              showError: false,
              checkOrder: true,
              ref: state.dateTimeRangePickerRef,
              expectedFormat: state.formatString,
              valueMode: "dateTimePicker",
              required: state.required
            }
          );
          return {
            dateTimeRangePickerTimeframe: timeframeWithPrecision,
            calendarValue: timeframeWithPrecision,
            calendarError: dateTimeRangePickerError,
            dateTimeRangePickerError
          };
        });
      },
      clearCalendarValue: () => {
        set((state) => {
          if (!state.calendarValue || state.calendarValue.from && state.calendarValue.to) {
            return state;
          }
          if (state.dateTimeRangePickerTimeframe?.from && state.dateTimeRangePickerTimeframe?.to) {
            return { calendarValue: state.dateTimeRangePickerTimeframe };
          }
          return { calendarValue: null };
        });
      }
    };
  };
  return actions;
}
export {
  getDateTimeRangePickerActions
};
