// @dynatrace/strato-components/esm/core/time/timeframeStore/actions/dateTimePicker.js
import { focusElementWithRetry } from "../../../../forms/date-time-picker/utils/focusDateTimeInput.js";
import {
  getTimeValueFromSimpleValue,
  toCalendarValue
} from "../../../../forms/forms-core/time/helpers.js";
import { getTimeValue } from "../../../../forms/forms-core/time/timeParser.js";
import { validate } from "../../validation.js";
function getDateTimePickerActions(formFieldMessagesDispatch, onChangeProp, isControlled) {
  const actions = (set) => {
    return {
      setCalendarValue: (inputValue) => {
        set((state) => {
          const newTimeValue = getTimeValueFromSimpleValue(
            inputValue,
            state.timeOffset,
            void 0,
            void 0,
            state.min,
            state.max,
            state.precision
          ) ?? null;
          const calendarError = validate(newTimeValue, state.min, state.max, {
            formMessageDispatch: formFieldMessagesDispatch,
            formerError: state.dateTimePickerError,
            precision: state.precision,
            showError: false,
            checkOrder: true,
            ref: state.dateTimePickerRef,
            expectedFormat: state.formatString,
            valueMode: "single",
            required: state.required
          });
          state.dateTimePickerRef?.current?.close();
          focusElementWithRetry(
            () => state.dateTimeInputRef.current?.spinButtonHandlers.get("hour")?.element.current,
            state.dateTimeInputRef
          );
          const calendarTimeValue = toCalendarValue(
            newTimeValue,
            state.precision
          );
          return {
            calendarValue: calendarTimeValue,
            dateTimePickerValue: newTimeValue,
            dateTimePickerError: calendarError,
            calendarError
          };
        });
      },
      setDateTimePickerValue: (value) => {
        return set((state) => {
          const newValue = getTimeValue(value, state.precision) ?? null;
          const isPartialValue = newValue?.absoluteDate === "";
          if (isPartialValue) {
            const errorKey = validate(newValue, state.min, state.max, {
              formMessageDispatch: formFieldMessagesDispatch,
              formerError: state.dateTimePickerError,
              precision: state.precision,
              showError: false,
              checkOrder: true,
              ref: state.dateTimePickerRef,
              expectedFormat: state.formatString,
              valueMode: "single"
            });
            onChangeProp?.("");
            return {
              calendarError: errorKey,
              dateTimePickerError: errorKey
            };
          }
          onChangeProp?.(newValue);
          const calendarTimeValue = toCalendarValue(newValue, state.precision);
          if (isControlled) {
            return { calendarValue: calendarTimeValue };
          }
          const dateTimePickerError = validate(newValue, state.min, state.max, {
            formMessageDispatch: formFieldMessagesDispatch,
            formerError: state.dateTimePickerError,
            precision: state.precision,
            showError: false,
            checkOrder: false,
            ref: state.dateTimePickerRef,
            expectedFormat: state.formatString,
            valueMode: "single"
          });
          return {
            dateTimePickerValue: newValue,
            calendarValue: calendarTimeValue,
            calendarError: dateTimePickerError,
            dateTimePickerError
          };
        });
      }
    };
  };
  return actions;
}
export {
  getDateTimePickerActions
};
