// @dynatrace/strato-components/esm/core/time/timeframeStore/actions/timeframeSelector.js
import { TZDateMini } from "@date-fns/tz";
import { endOfDay, startOfDay } from "date-fns";
import { getTimezone } from "@dynatrace-sdk/user-preferences";
import { getTimeframeFromCalendarSelection } from "./calendar.js";
import { focusElementWithRetry } from "../../../../forms/date-time-picker/utils/focusDateTimeInput.js";
import {
  removePrecision,
  removePrecisionFromTimeFrame
} from "../../../../forms/forms-core/time/timeParser.js";
import {
  isFullTimeframe,
  isTimeValue,
  mergeRangeValue
} from "../../helpers.js";
import { validate } from "../../validation.js";
function handleTimeframeSelectorTimeframeChange(inputValue, state, isControlled, timeframeSelectorRef, formFieldMessagesDispatch, onChangeProp, errorMessages) {
  const timeframeWithPrecision = removePrecisionFromTimeFrame(inputValue, state.precision) ?? null;
  const value = isFullTimeframe(timeframeWithPrecision) ? timeframeWithPrecision : null;
  onChangeProp?.(value);
  timeframeSelectorRef.current?.close();
  if (isControlled) {
    return state;
  }
  const error = validate(timeframeWithPrecision, state.min, state.max, {
    formMessageDispatch: formFieldMessagesDispatch,
    formerError: state.timeframeSelectorError,
    precision: state.precision,
    showError: false,
    checkOrder: true,
    ref: timeframeSelectorRef,
    expectedFormat: state.formatString,
    valueMode: "timeframeSelector",
    required: state.required,
    errorMessages
  });
  return {
    timeframeSelectorTimeframe: value,
    calendarValue: value ?? timeframeWithPrecision,
    dateTimeRangePickerTimeframe: value ?? timeframeWithPrecision,
    timeframeSelectorError: error,
    calendarError: error,
    dateTimeRangePickerError: error
  };
}
function handlePreviousTimeframeChange(value) {
  return {
    referenceTimeframe: value
  };
}
function isFullDayTimeframe(timeframe, state) {
  const timeZone = getTimezone();
  const to = new TZDateMini(timeframe.to.absoluteDate, timeZone);
  const from = new TZDateMini(timeframe.from.absoluteDate, timeZone);
  const isToAtEndOfDay = removePrecision(endOfDay(to).toISOString(), state.precision) === removePrecision(timeframe.to.absoluteDate, state.precision);
  const isToAtMax = state.max?.absoluteDate === to.toISOString();
  const isFromAtStartOfDay = removePrecision(startOfDay(from).toISOString(), state.precision) === removePrecision(timeframe.from.absoluteDate, state.precision);
  const isFromAtMin = state.min?.absoluteDate === from.toISOString();
  return Boolean(
    (isToAtEndOfDay || isToAtMax) && (isFromAtStartOfDay || isFromAtMin)
  );
}
function getTimeframeSelectorActions(isControlled, timeframeSelectorRef, formFieldMessagesDispatch, onChangeProp, debouncedOnChangeRef, errorMessages) {
  const actions = (set) => {
    return {
      setCalendarValue: (inputValue) => {
        set((state) => {
          const calendarValue = getTimeframeFromCalendarSelection(
            inputValue,
            state
          );
          if (isFullTimeframe(calendarValue)) {
            state.dateTimeRangePickerRef?.current?.close();
            const isFullDay = isFullDayTimeframe(calendarValue, state);
            focusElementWithRetry(
              () => state.dateTimeFromRef.current?.spinButtonHandlers.get("hour")?.element.current,
              state.dateTimeFromRef
            );
            return {
              calendarValue,
              dateTimeRangePickerTimeframe: calendarValue,
              isFullDay
            };
          }
          return { calendarValue, isFullDay: false };
        });
      },
      setDateTimeRangePickerTimeframe: (value) => {
        return set((state) => {
          if (value === null) {
            return {
              dateTimeRangePickerTimeframe: null,
              calendarValue: null,
              dateTimeRangePickerError: void 0,
              calendarError: void 0
            };
          }
          const merged = mergeRangeValue(
            state.dateTimeRangePickerTimeframe,
            value
          );
          const timeframeWithPrecision = removePrecisionFromTimeFrame(merged, state.precision) ?? null;
          const isPartialFrom = isTimeValue(merged.from) && merged.from.absoluteDate === "";
          const isPartialTo = isTimeValue(merged.to) && merged.to.absoluteDate === "";
          if (isPartialFrom || isPartialTo) {
            const errorKey = validate(
              timeframeWithPrecision,
              state.min,
              state.max,
              {
                formerError: state.dateTimeRangePickerError,
                precision: state.precision,
                showError: false,
                checkOrder: true,
                ref: state.dateTimeRangePickerRef,
                expectedFormat: state.formatString,
                valueMode: "timeframeSelector",
                required: state.required,
                errorMessages
              }
            );
            const calendarValue = {
              from: isPartialFrom ? null : timeframeWithPrecision?.from ?? null,
              to: isPartialTo ? null : timeframeWithPrecision?.to ?? null
            };
            return {
              dateTimeRangePickerTimeframe: timeframeWithPrecision,
              calendarValue,
              calendarError: errorKey,
              dateTimeRangePickerError: errorKey
            };
          }
          const dateTimeRangePickerError = validate(
            timeframeWithPrecision,
            state.min,
            state.max,
            {
              formerError: state.dateTimeRangePickerError,
              precision: state.precision,
              showError: false,
              checkOrder: true,
              ref: state.dateTimeRangePickerRef,
              expectedFormat: state.formatString,
              valueMode: "timeframeSelector",
              required: state.required,
              errorMessages
            }
          );
          return {
            dateTimeRangePickerTimeframe: timeframeWithPrecision,
            calendarValue: timeframeWithPrecision,
            calendarError: dateTimeRangePickerError,
            dateTimeRangePickerError
          };
        });
      },
      setTimeframeSelectorTimeframe: (inputValue) => {
        debouncedOnChangeRef?.current?.cancel();
        return set(
          (state) => handleTimeframeSelectorTimeframeChange(
            inputValue,
            state,
            isControlled,
            timeframeSelectorRef,
            formFieldMessagesDispatch,
            onChangeProp,
            errorMessages
          )
        );
      },
      setReferenceTimeframe: (value) => {
        set(() => handlePreviousTimeframeChange(value));
      },
      setStepperValue: (value) => {
        set(() => {
          debouncedOnChangeRef?.current?.(value);
          return {
            timeframeSelectorTimeframe: value,
            calendarValue: value,
            dateTimeRangePickerTimeframe: value
          };
        });
      },
      clearCalendarValue: () => {
        set((state) => {
          if (!state.calendarValue || state.calendarValue.from && state.calendarValue.to) {
            return state;
          }
          if (state.dateTimeRangePickerTimeframe?.from && state.dateTimeRangePickerTimeframe?.to) {
            return { calendarValue: state.dateTimeRangePickerTimeframe };
          }
          return { calendarValue: null };
        });
      }
    };
  };
  return actions;
}
export {
  getTimeframeSelectorActions
};
