// @dynatrace/strato-components/charts/core/components/text-ellipsis/chars-font-width-map.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var chars_font_width_map_exports = {};
__export(chars_font_width_map_exports, {
  charsFontWidth12pixelsSize400Map: () => charsFontWidth12pixelsSize400Map,
  charsFontWidth12pixelsSize500Map: () => charsFontWidth12pixelsSize500Map,
  charsFontWidth12pixelsSize600Map: () => charsFontWidth12pixelsSize600Map,
  charsFontWidth12pixelsSize700Map: () => charsFontWidth12pixelsSize700Map,
  charsFontWidth14pixelsSize400Map: () => charsFontWidth14pixelsSize400Map,
  charsFontWidth16pixelsSize500Map: () => charsFontWidth16pixelsSize500Map,
  charsFontWidth9pixelsSize500Map: () => charsFontWidth9pixelsSize500Map
});
module.exports = __toCommonJS(chars_font_width_map_exports);
const charsFontWidth12pixelsSize400Map = {
  "0": 6.8879852294921875,
  "1": 6.8879852294921875,
  "2": 6.8879852294921875,
  "3": 6.8879852294921875,
  "4": 6.8879852294921875,
  "5": 6.8879852294921875,
  "6": 6.8879852294921875,
  "7": 6.8879852294921875,
  "8": 6.8879852294921875,
  "9": 6.8879852294921875,
  a: 6.5879974365234375,
  \u00E1: 6.5879974365234375,
  \u00E0: 6.5879974365234375,
  \u00E4: 6.5879974365234375,
  \u00E2: 6.5879974365234375,
  \u00E3: 6.5879974365234375,
  \u00E5: 6.5879974365234375,
  b: 6.5999908447265625,
  c: 5.7719879150390625,
  \u00E7: 5.7719879150390625,
  d: 6.5999908447265625,
  e: 6.29998779296875,
  \u00E9: 6.29998779296875,
  \u00E8: 6.29998779296875,
  \u00EB: 6.29998779296875,
  \u00EA: 6.29998779296875,
  f: 4.391998291015625,
  g: 6.5279998779296875,
  h: 6.85198974609375,
  i: 3.02398681640625,
  \u00ED: 3.02398681640625,
  \u00EC: 3.02398681640625,
  \u00EF: 3.02398681640625,
  j: 3.1079864501953125,
  k: 6.0479888916015625,
  l: 3.39599609375,
  m: 10.727996826171875,
  n: 6.839996337890625,
  \u00F1: 6.839996337890625,
  o: 6.33599853515625,
  \u00F3: 6.33599853515625,
  \u00F2: 6.33599853515625,
  \u00F6: 6.33599853515625,
  \u00F4: 6.33599853515625,
  \u00F5: 6.33599853515625,
  \u00F8: 6.33599853515625,
  p: 6.5999908447265625,
  q: 6.61199951171875,
  r: 4.7159881591796875,
  s: 5.8079986572265625,
  \u0161: 5.8079986572265625,
  t: 4.4399871826171875,
  u: 6.839996337890625,
  \u00FA: 6.839996337890625,
  \u00F9: 6.839996337890625,
  \u00FC: 6.839996337890625,
  v: 5.855987548828125,
  w: 8.855987548828125,
  x: 5.9879913330078125,
  y: 5.7239990234375,
  \u00FD: 5.7239990234375,
  \u00FF: 5.7239990234375,
  z: 5.9639892578125,
  A: 7.5959930419921875,
  \u00C1: 7.5959930419921875,
  \u00C0: 7.5959930419921875,
  \u00C4: 7.5959930419921875,
  \u00C2: 7.5959930419921875,
  \u00C3: 7.5959930419921875,
  \u00C5: 7.5959930419921875,
  B: 7.4639892578125,
  C: 7.1639862060546875,
  \u00C7: 7.1639862060546875,
  D: 7.4759979248046875,
  E: 6.947998046875,
  \u00C9: 6.947998046875,
  \u00C8: 6.947998046875,
  \u00CB: 6.947998046875,
  \u00CA: 6.947998046875,
  F: 6.5879974365234375,
  G: 7.4879913330078125,
  H: 8.147994995117188,
  I: 4.4759979248046875,
  \u00CD: 4.4759979248046875,
  \u00CC: 4.4759979248046875,
  \u00CF: 4.4759979248046875,
  J: 5.4239959716796875,
  K: 7.28399658203125,
  L: 6.239990234375,
  M: 9.7919921875,
  N: 8.21998596191406,
  \u00D1: 8.21998596191406,
  O: 7.67999267578125,
  \u00D3: 7.67999267578125,
  \u00D2: 7.67999267578125,
  \u00D6: 7.67999267578125,
  \u00D4: 7.67999267578125,
  \u00D5: 7.67999267578125,
  \u00D8: 7.67999267578125,
  P: 7.1999969482421875,
  Q: 7.67999267578125,
  R: 7.6079864501953125,
  S: 7.0919952392578125,
  \u0160: 7.0919952392578125,
  T: 6.4799957275390625,
  U: 7.8119964599609375,
  \u00DA: 7.8119964599609375,
  \u00D9: 7.8119964599609375,
  \u00DC: 7.8119964599609375,
  V: 7.4759979248046875,
  W: 11.327987670898438,
  X: 7.83599853515625,
  Y: 7.115997314453125,
  \u00DD: 7.115997314453125,
  \u0178: 7.115997314453125,
  Z: 7.24798583984375,
  " ": 2.9759979248046875,
  "!": 3.45599365234375,
  '"': 4.4639892578125,
  "#": 7.631988525390625,
  $: 6.899993896484375,
  "%": 8.855987548828125,
  "&": 7.6439971923828125,
  "'": 2.62799072265625,
  "(": 4.1039886474609375,
  ")": 3.8879852294921875,
  "*": 6.1439971923828125,
  "+": 6.8879852294921875,
  ",": 2.9759979248046875,
  "-": 5.6639862060546875,
  ".": 2.9759979248046875,
  "/": 4.1039886474609375,
  ":": 2.9759979248046875,
  ";": 2.9759979248046875,
  "<": 7.2119903564453125,
  "=": 6.8879852294921875,
  ">": 7.2119903564453125,
  "?": 6.5519866943359375,
  "@": 12.7919921875,
  "[": 4.055999755859375,
  "\\": 4.1039886474609375,
  "]": 3.9839935302734375,
  "^": 6.8879852294921875,
  _: 6.8159942626953125,
  "`": 3.29998779296875,
  "{": 4.4279937744140625,
  "|": 3.1199951171875,
  "}": 4.4279937744140625,
  "~": 6.8879852294921875,
  "\u2022": 5.23199462890625,
  "\u2026": 8.927993774414062,
  "\xB0": 4.511993408203125,
  \u00B5: 6.839996337890625,
  "\xB2": 4.12799072265625,
  "\xB3": 4.12799072265625,
  "\xB1": 6.8879852294921875,
  "\u20AC": 6.899993896484375,
  "\xA3": 6.899993896484375,
  "\xA5": 6.8879852294921875
};
const charsFontWidth14pixelsSize400Map = {
  "0": 8.035995483398438,
  "1": 8.035995483398438,
  "2": 8.035995483398438,
  "3": 8.035995483398438,
  "4": 8.035995483398438,
  "5": 8.035995483398438,
  "6": 8.035995483398438,
  "7": 8.035995483398438,
  "8": 8.035995483398438,
  "9": 8.035995483398438,
  a: 7.6859893798828125,
  b: 7.6999969482421875,
  c: 6.7339935302734375,
  d: 7.6999969482421875,
  e: 7.3499908447265625,
  f: 5.123992919921875,
  g: 7.615997314453125,
  h: 7.993988037109375,
  i: 3.5279998779296875,
  j: 3.6259918212890625,
  k: 7.055999755859375,
  l: 3.9619903564453125,
  m: 12.5159912109375,
  n: 7.9799957275390625,
  o: 7.391998291015625,
  p: 7.6999969482421875,
  q: 7.7139892578125,
  r: 5.5019989013671875,
  s: 6.7759857177734375,
  t: 5.17999267578125,
  u: 7.9799957275390625,
  v: 6.8319854736328125,
  w: 10.33198547363281,
  x: 6.985992431640625,
  y: 6.6779937744140625,
  z: 6.9579925537109375,
  \u00C1: 8.86199951171875,
  \u00C0: 8.86199951171875,
  \u00C4: 8.86199951171875,
  \u00C2: 8.86199951171875,
  \u00C3: 8.86199951171875,
  \u00C5: 8.86199951171875,
  \u00C7: 8.35798645019531,
  \u00C9: 8.105987548828125,
  \u00C8: 8.105987548828125,
  \u00CB: 8.105987548828125,
  \u00CA: 8.105987548828125,
  \u00CD: 5.2220001220703125,
  \u00CC: 5.2220001220703125,
  \u00CF: 5.2220001220703125,
  \u00D1: 9.589996337890625,
  \u00D3: 8.959991455078125,
  \u00D2: 8.959991455078125,
  \u00D6: 8.959991455078125,
  \u00D4: 8.959991455078125,
  \u00D5: 8.959991455078125,
  \u00D8: 8.959991455078125,
  \u0160: 8.27398681640625,
  \u00DD: 8.301986694335938,
  \u0178: 8.301986694335938,
  \u00E1: 7.6859893798828125,
  \u00E0: 7.6859893798828125,
  \u00E4: 7.6859893798828125,
  \u00E2: 7.6859893798828125,
  \u00E3: 7.6859893798828125,
  \u00E5: 7.6859893798828125,
  \u00E7: 6.7339935302734375,
  \u00E9: 7.3499908447265625,
  \u00E8: 7.3499908447265625,
  \u00EB: 7.3499908447265625,
  \u00EA: 7.3499908447265625,
  \u00ED: 3.5279998779296875,
  \u00EC: 3.5279998779296875,
  \u00EF: 3.5279998779296875,
  \u00F1: 7.9799957275390625,
  \u00F3: 7.391998291015625,
  \u00F2: 7.391998291015625,
  \u00F6: 7.391998291015625,
  \u00F4: 7.391998291015625,
  \u00F5: 7.391998291015625,
  \u00F8: 7.391998291015625,
  \u0161: 6.7759857177734375,
  \u00FD: 6.6779937744140625,
  \u00FF: 6.6779937744140625,
  A: 8.86199951171875,
  B: 8.707992553710938,
  C: 8.35798645019531,
  D: 8.72200012207031,
  E: 8.105987548828125,
  F: 7.6859893798828125,
  G: 8.735992431640625,
  H: 9.50599670410156,
  I: 5.2220001220703125,
  J: 6.3279876708984375,
  K: 8.49798583984375,
  L: 7.279998779296875,
  M: 11.423995971679688,
  N: 9.589996337890625,
  O: 8.959991455078125,
  P: 8.399993896484375,
  Q: 8.959991455078125,
  R: 8.87599182128906,
  S: 8.27398681640625,
  T: 7.55999755859375,
  U: 9.113998413085938,
  V: 8.72200012207031,
  W: 13.215988159179688,
  X: 9.141998291015625,
  Y: 8.301986694335938,
  Z: 8.45599365234375,
  " ": 3.47198486328125,
  "!": 4.0319976806640625,
  '"': 5.2079925537109375,
  "'": 3.0659942626953125,
  "#": 8.90399169921875,
  $: 8.04998779296875,
  "%": 10.33198547363281,
  "&": 8.917999267578125,
  "(": 4.787994384765625,
  ")": 4.5359954833984375,
  "*": 7.167999267578125,
  "+": 8.035995483398438,
  ",": 3.47198486328125,
  "-": 6.6079864501953125,
  ".": 3.47198486328125,
  "/": 4.787994384765625,
  ":": 3.47198486328125,
  ";": 3.47198486328125,
  "<": 8.413986206054688,
  "=": 8.035995483398438,
  ">": 8.413986206054688,
  "?": 7.6439971923828125,
  "@": 14.923995971679688,
  "[": 4.73199462890625,
  "\\": 4.787994384765625,
  "]": 4.6479949951171875,
  "^": 8.035995483398438,
  _: 7.951995849609375,
  "`": 3.8499908447265625,
  "{": 5.165985107421875,
  "|": 3.6399993896484375,
  "}": 5.165985107421875,
  "~": 8.035995483398438,
  "\u2022": 6.1039886474609375,
  "\u2026": 10.415985107421875,
  "\xB0": 5.2639923095703125,
  \u00B5: 7.9799957275390625,
  "\xB2": 4.8159942626953125,
  "\xB3": 4.8159942626953125,
  "\xB1": 8.035995483398438,
  "\u20AC": 8.04998779296875,
  "\xA3": 8.04998779296875,
  "\xA5": 8.035995483398438
};
const charsFontWidth12pixelsSize500Map = {
  "0": 6.959991,
  "1": 6.959991,
  "2": 6.959991,
  "3": 6.959991,
  "4": 6.959991,
  "5": 6.959991,
  "6": 6.959991,
  "7": 6.959991,
  "8": 6.959991,
  "9": 6.959991,
  a: 6.695999,
  b: 6.743988,
  c: 5.843994,
  d: 6.743988,
  e: 6.40799,
  f: 4.584,
  g: 6.68399,
  h: 6.972,
  i: 3.179993,
  j: 3.275986,
  k: 6.311996,
  l: 3.551987,
  m: 10.715988,
  n: 6.947998,
  o: 6.491989,
  p: 6.743988,
  q: 6.755997,
  r: 4.931992,
  s: 5.939987,
  t: 4.619995,
  u: 6.947998,
  v: 6.047989,
  w: 9.203995,
  x: 6.299988,
  y: 6.011993,
  z: 6.071991,
  \u00E1: 6.695999,
  \u00E0: 6.695999,
  \u00E4: 6.695999,
  \u00E2: 6.695999,
  \u00E3: 6.695999,
  \u00E5: 6.695999,
  \u00E7: 5.843994,
  \u00E9: 6.40799,
  \u00E8: 6.40799,
  \u00EB: 6.40799,
  \u00EA: 6.40799,
  \u00ED: 3.179993,
  \u00EC: 3.179993,
  \u00EF: 3.179993,
  \u00F1: 6.947998,
  \u00F3: 6.491989,
  \u00F2: 6.491989,
  \u00F6: 6.491989,
  \u00F4: 6.491989,
  \u00F5: 6.491989,
  \u00F8: 6.491989,
  \u0161: 5.939987,
  \u00FD: 6.011993,
  \u00FF: 6.011993,
  Z: 7.29599,
  A: 7.73999,
  \u00C1: 7.73999,
  \u00C0: 7.73999,
  \u00C4: 7.73999,
  \u00C2: 7.73999,
  \u00C3: 7.73999,
  \u00C5: 7.73999,
  B: 7.5,
  C: 7.247986,
  \u00C7: 7.247986,
  D: 7.643997,
  E: 7.056,
  \u00C9: 7.056,
  \u00C8: 7.056,
  \u00CB: 7.056,
  \u00CA: 7.056,
  F: 6.707993,
  G: 7.584,
  H: 8.291992,
  I: 4.631989,
  \u00CD: 4.631989,
  \u00CC: 4.631989,
  \u00CF: 4.631989,
  J: 5.543991,
  K: 7.5,
  L: 6.299988,
  " ": 3.071991,
  M: 10.127991,
  N: 8.399994,
  O: 7.835999,
  \u00D3: 7.835999,
  \u00D2: 7.835999,
  \u00D6: 7.835999,
  \u00D4: 7.835999,
  \u00D5: 7.835999,
  \u00D8: 7.835999,
  P: 7.331985,
  Q: 7.835999,
  R: 7.691986,
  S: 7.271988,
  \u0160: 7.271988,
  T: 6.575989,
  U: 7.955994,
  V: 7.655991,
  W: 11.639999,
  X: 8.003998,
  Y: 7.235992,
  \u00DD: 7.235992,
  \u0178: 7.235992,
  "!": 3.515991,
  '"': 4.73999,
  "'": 2.771988,
  "#": 7.631989,
  $: 7.019989,
  "%": 8.79599,
  "&": 7.547989,
  "(": 4.271988,
  ")": 4.127991,
  "*": 6.395996,
  "+": 6.899994,
  ",": 3.084,
  "-": 5.723999,
  ".": 3.084,
  "/": 4.367996,
  ":": 3.084,
  ";": 3.084,
  "<": 7.103989,
  "=": 6.899994,
  ">": 7.103989,
  "?": 6.659988,
  "@": 12.791992,
  "[": 4.37999,
  "\\": 4.367996,
  "]": 4.319992,
  "^": 6.899994,
  _: 6.959991,
  "`": 3.515991,
  "{": 4.691986,
  "|": 3.335999,
  "}": 4.691986,
  "~": 6.899994,
  "\u2022": 5.255997,
  "\u2026": 9.251991,
  "\xB0": 4.583984,
  \u00B5: 6.947998,
  "\xB2": 4.175995,
  "\xB3": 4.175995,
  "\xB1": 6.899994,
  "\u20AC": 7.019989,
  "\xA3": 7.019989,
  "\xA5": 6.899994
};
const charsFontWidth12pixelsSize600Map = {
  "0": 7.031998,
  "1": 7.031998,
  "2": 7.031998,
  "3": 7.031998,
  "4": 7.031998,
  "5": 7.031998,
  "6": 7.031998,
  "7": 7.031998,
  "8": 7.031998,
  "9": 7.031998,
  a: 6.815994,
  b: 6.911987,
  c: 5.927994,
  d: 6.911987,
  e: 6.539993,
  f: 4.811996,
  g: 6.875992,
  h: 7.115997,
  i: 3.359985,
  j: 3.479996,
  k: 6.623993,
  l: 3.743988,
  m: 10.691986,
  n: 7.079987,
  o: 6.659988,
  p: 6.911987,
  q: 6.923996,
  r: 5.18399,
  s: 6.095993,
  t: 4.82399,
  u: 7.079987,
  v: 6.275986,
  w: 9.612,
  x: 6.671997,
  y: 6.347992,
  z: 6.191986,
  \u00E1: 6.815994,
  \u00E0: 6.815994,
  \u00E4: 6.815994,
  \u00E2: 6.815994,
  \u00E3: 6.815994,
  \u00E5: 6.815994,
  \u00E7: 5.927994,
  \u00E9: 6.539993,
  \u00E8: 6.539993,
  \u00EB: 6.539993,
  \u00EA: 6.539993,
  \u00ED: 3.359985,
  \u00EC: 3.359985,
  \u00EF: 3.359985,
  \u00F1: 7.079987,
  \u00F3: 6.659988,
  \u00F2: 6.659988,
  \u00F6: 6.659988,
  \u00F4: 6.659988,
  \u00F5: 6.659988,
  \u00F8: 6.659988,
  \u0161: 6.095993,
  \u00FD: 6.347992,
  \u00FF: 6.347992,
  Z: 7.367996,
  A: 7.895996,
  \u00C1: 7.895996,
  \u00C0: 7.895996,
  \u00C4: 7.895996,
  \u00C2: 7.895996,
  \u00C3: 7.895996,
  \u00C5: 7.895996,
  B: 7.547989,
  C: 7.343994,
  \u00C7: 7.343994,
  D: 7.82399,
  E: 7.175995,
  \u00C9: 7.175995,
  \u00C8: 7.175995,
  \u00CB: 7.175995,
  \u00CA: 7.175995,
  F: 6.85199,
  G: 7.703995,
  H: 8.447998,
  I: 4.811996,
  \u00CD: 4.811996,
  \u00CC: 4.811996,
  \u00CF: 4.811996,
  J: 5.687988,
  K: 7.73999,
  L: 6.371994,
  " ": 3.167999,
  M: 10.523987,
  N: 8.615997,
  O: 8.015991,
  \u00D3: 8.015991,
  \u00D2: 8.015991,
  \u00D6: 8.015991,
  \u00D4: 8.015991,
  \u00D5: 8.015991,
  \u00D8: 8.015991,
  P: 7.487991,
  Q: 8.015991,
  R: 7.787994,
  S: 7.475998,
  \u0160: 7.475998,
  T: 6.68399,
  U: 8.112,
  V: 7.859985,
  W: 11.987991,
  X: 8.195999,
  Y: 7.367996,
  \u00DD: 7.367996,
  \u0178: 7.367996,
  "!": 3.587997,
  '"': 5.063995,
  "'": 2.951996,
  "#": 7.631989,
  $: 7.163986,
  "%": 8.735992,
  "&": 7.427994,
  "(": 4.463989,
  ")": 4.391998,
  "*": 6.695999,
  "+": 6.923996,
  ",": 3.215988,
  "-": 5.79599,
  ".": 3.215988,
  "/": 4.679993,
  ":": 3.215988,
  ";": 3.215988,
  "<": 6.983994,
  "=": 6.923996,
  ">": 6.983994,
  "?": 6.791992,
  "@": 12.791992,
  "[": 4.763992,
  "\\": 4.679993,
  "]": 4.703995,
  "^": 6.923996,
  _: 7.115997,
  "`": 3.76799,
  "{": 5.003998,
  "|": 3.575989,
  "}": 5.003998,
  "~": 6.923996,
  "\u2022": 5.291992,
  "\u2026": 9.647988,
  "\xB0": 4.655991,
  \u00B5: 7.079987,
  "\xB2": 4.223999,
  "\xB3": 4.223999,
  "\xB1": 6.923996,
  "\u20AC": 7.163986,
  "\xA3": 7.163986,
  "\xA5": 6.923996
};
const charsFontWidth12pixelsSize700Map = {
  "0": 6.673828,
  "1": 6.673828,
  "2": 6.673828,
  "3": 6.673828,
  "4": 6.673828,
  "5": 6.673828,
  "6": 6.673828,
  "7": 6.673828,
  "8": 6.673828,
  "9": 6.673828,
  a: 6.673828,
  b: 7.330078,
  c: 6.673828,
  d: 7.330078,
  e: 6.673828,
  f: 3.996094,
  g: 7.330078,
  h: 7.330078,
  i: 3.333984,
  j: 3.333984,
  k: 6.673828,
  l: 3.333984,
  m: 10.669922,
  n: 7.330078,
  o: 7.330078,
  p: 7.330078,
  q: 7.330078,
  r: 4.669922,
  s: 6.673828,
  t: 3.996094,
  u: 7.330078,
  v: 6.673828,
  w: 9.333984,
  x: 6.673828,
  y: 6.673828,
  z: 6,
  \u00E1: 6.673828,
  \u00E0: 6.673828,
  \u00E4: 6.673828,
  \u00E2: 6.673828,
  \u00E3: 6.673828,
  \u00E5: 6.673828,
  \u00E7: 6.673828,
  \u00E9: 6.673828,
  \u00E8: 6.673828,
  \u00EB: 6.673828,
  \u00EA: 6.673828,
  \u00ED: 3.333984,
  \u00EC: 3.333984,
  \u00EF: 3.333984,
  \u00F1: 7.330078,
  \u00F3: 7.330078,
  \u00F2: 7.330078,
  \u00F6: 7.330078,
  \u00F4: 7.330078,
  \u00F5: 7.330078,
  \u00F8: 7.330078,
  \u0161: 6.673828,
  \u00FD: 6.673828,
  \u00FF: 6.673828,
  Z: 7.330078,
  A: 8.666016,
  \u00C1: 8.666016,
  \u00C0: 8.666016,
  \u00C4: 8.666016,
  \u00C2: 8.666016,
  \u00C3: 8.666016,
  \u00C5: 8.666016,
  B: 8.666016,
  C: 8.666016,
  \u00C7: 8.666016,
  D: 8.666016,
  E: 8.003906,
  \u00C9: 8.003906,
  \u00C8: 8.003906,
  \u00CB: 8.003906,
  \u00CA: 8.003906,
  F: 7.330078,
  G: 9.333984,
  H: 8.666016,
  I: 3.333984,
  \u00CD: 3.333984,
  \u00CC: 3.333984,
  \u00CF: 3.333984,
  J: 6.673828,
  K: 8.666016,
  L: 7.330078,
  " ": 3.333984,
  M: 9.996094,
  N: 8.666016,
  O: 9.333984,
  \u00D3: 9.333984,
  \u00D2: 9.333984,
  \u00D6: 9.333984,
  \u00D4: 9.333984,
  \u00D5: 9.333984,
  \u00D8: 9.333984,
  P: 8.003906,
  Q: 9.333984,
  R: 8.666016,
  S: 8.003906,
  \u0160: 8.003906,
  T: 7.330078,
  U: 8.666016,
  V: 8.003906,
  W: 11.326172,
  X: 8.003906,
  Y: 8.003906,
  \u00DD: 8.003906,
  \u0178: 8.003906,
  "!": 3.996094,
  '"': 5.689453,
  "'": 2.853516,
  "#": 6.673828,
  $: 6.673828,
  "%": 10.669922,
  "&": 8.666016,
  "(": 3.996094,
  ")": 3.996094,
  "*": 4.669922,
  "+": 7.007813,
  ",": 3.333984,
  "-": 3.996094,
  ".": 3.333984,
  "/": 3.333984,
  ":": 3.996094,
  ";": 3.996094,
  "<": 7.007813,
  "=": 7.007813,
  ">": 7.007813,
  "?": 7.330078,
  "@": 11.701172,
  "[": 3.996094,
  "\\": 3.333984,
  "]": 3.996094,
  "^": 7.007813,
  _: 6.673828,
  "`": 3.996094,
  "{": 4.669922,
  "|": 3.357422,
  "}": 4.669922,
  "~": 7.007813,
  "\u2022": 4.201172,
  "\u2026": 10.001953,
  "\xB0": 4.669922,
  \u00B5: 7.330078,
  "\xB2": 4.001953,
  "\xB3": 4.001953,
  "\xB1": 7.007813,
  "\u20AC": 6.673828,
  "\xA3": 6.673828,
  "\xA5": 7.007813
};
const charsFontWidth9pixelsSize500Map = {
  "0": 5.219993,
  "1": 5.219993,
  "2": 5.219993,
  "3": 5.219993,
  "4": 5.219993,
  "5": 5.219993,
  "6": 5.219993,
  "7": 5.219993,
  "8": 5.219993,
  "9": 5.219993,
  a: 5.021999,
  b: 5.057991,
  c: 4.382995,
  d: 5.057991,
  e: 4.805993,
  f: 3.438,
  g: 5.012992,
  h: 5.229,
  i: 2.384995,
  j: 2.45699,
  k: 4.733997,
  l: 2.66399,
  m: 8.036991,
  n: 5.210999,
  o: 4.868992,
  p: 5.057991,
  q: 5.066998,
  r: 3.698994,
  s: 4.45499,
  t: 3.464996,
  u: 5.210999,
  v: 4.535992,
  w: 6.902996,
  x: 4.724991,
  y: 4.508995,
  z: 4.553993,
  \u00E1: 5.021999,
  \u00E0: 5.021999,
  \u00E4: 5.021999,
  \u00E2: 5.021999,
  \u00E3: 5.021999,
  \u00E5: 5.021999,
  \u00E7: 4.382995,
  \u00E9: 4.805993,
  \u00E8: 4.805993,
  \u00EB: 4.805993,
  \u00EA: 4.805993,
  \u00ED: 2.384995,
  \u00EC: 2.384995,
  \u00EF: 2.384995,
  \u00F1: 5.210999,
  \u00F3: 4.868992,
  \u00F2: 4.868992,
  \u00F6: 4.868992,
  \u00F4: 4.868992,
  \u00F5: 4.868992,
  \u00F8: 4.868992,
  \u0161: 4.45499,
  \u00FD: 4.508995,
  \u00FF: 4.508995,
  Z: 5.471992,
  A: 5.804992,
  \u00C1: 5.804992,
  \u00C0: 5.804992,
  \u00C4: 5.804992,
  \u00C2: 5.804992,
  \u00C3: 5.804992,
  \u00C5: 5.804992,
  B: 5.625,
  C: 5.435989,
  \u00C7: 5.435989,
  D: 5.732998,
  E: 5.292,
  \u00C9: 5.292,
  \u00C8: 5.292,
  \u00CB: 5.292,
  \u00CA: 5.292,
  F: 5.030995,
  G: 5.688,
  H: 6.218994,
  I: 3.473992,
  \u00CD: 3.473992,
  \u00CC: 3.473992,
  \u00CF: 3.473992,
  J: 4.157993,
  K: 5.625,
  L: 4.724991,
  " ": 2.303993,
  M: 7.595993,
  N: 6.299995,
  O: 5.876999,
  \u00D3: 5.876999,
  \u00D2: 5.876999,
  \u00D6: 5.876999,
  \u00D4: 5.876999,
  \u00D5: 5.876999,
  \u00D8: 5.876999,
  P: 5.498989,
  Q: 5.876999,
  R: 5.768989,
  S: 5.453991,
  \u0160: 5.453991,
  T: 4.931992,
  U: 5.966995,
  V: 5.741993,
  W: 8.729999,
  X: 6.002998,
  Y: 5.426994,
  \u00DD: 5.426994,
  \u0178: 5.426994,
  "!": 2.636993,
  '"': 3.554992,
  "'": 2.078991,
  "#": 5.723992,
  $: 5.264992,
  "%": 6.596992,
  "&": 5.660992,
  "(": 3.203991,
  ")": 3.095993,
  "*": 4.796997,
  "+": 5.174996,
  ",": 2.313,
  "-": 4.292999,
  ".": 2.313,
  "/": 3.275997,
  ":": 2.313,
  ";": 2.313,
  "<": 5.327992,
  "=": 5.174996,
  ">": 5.327992,
  "?": 4.994991,
  "@": 9.593994,
  "[": 3.284993,
  "\\": 3.275997,
  "]": 3.239994,
  "^": 5.174996,
  _: 5.219993,
  "`": 2.636993,
  "{": 3.518989,
  "|": 2.501999,
  "}": 3.518989,
  "~": 5.174996,
  "\u2022": 3.941998,
  "\u2026": 6.938993,
  "\xB0": 3.437988,
  \u00B5: 5.210999,
  "\xB2": 3.131996,
  "\xB3": 3.131996,
  "\xB1": 5.174996,
  "\u20AC": 5.264992,
  "\xA3": 5.264992,
  "\xA5": 5.174996
};
const charsFontWidth16pixelsSize500Map = {
  "0": 9.279988,
  "1": 9.279988,
  "2": 9.279988,
  "3": 9.279988,
  "4": 9.279988,
  "5": 9.279988,
  "6": 9.279988,
  "7": 9.279988,
  "8": 9.279988,
  "9": 9.279988,
  a: 8.927999,
  b: 8.991984,
  c: 7.791992,
  d: 8.991984,
  e: 8.543987,
  f: 6.112,
  g: 8.911987,
  h: 9.296,
  i: 4.239991,
  j: 4.367981,
  k: 8.415995,
  l: 4.735983,
  m: 14.287984,
  n: 9.263997,
  o: 8.655985,
  p: 8.991984,
  q: 9.007996,
  r: 6.575989,
  s: 7.919983,
  t: 6.159993,
  u: 9.263997,
  v: 8.063985,
  w: 12.271993,
  x: 8.399984,
  y: 8.015991,
  z: 8.095988,
  \u00E1: 8.927999,
  \u00E0: 8.927999,
  \u00E4: 8.927999,
  \u00E2: 8.927999,
  \u00E3: 8.927999,
  \u00E5: 8.927999,
  \u00E7: 7.791992,
  \u00E9: 8.543987,
  \u00E8: 8.543987,
  \u00EB: 8.543987,
  \u00EA: 8.543987,
  \u00ED: 4.239991,
  \u00EC: 4.239991,
  \u00EF: 4.239991,
  \u00F1: 9.263997,
  \u00F3: 8.655985,
  \u00F2: 8.655985,
  \u00F6: 8.655985,
  \u00F4: 8.655985,
  \u00F5: 8.655985,
  \u00F8: 8.655985,
  \u0161: 7.919983,
  \u00FD: 8.015991,
  \u00FF: 8.015991,
  Z: 9.727987,
  A: 10.319987,
  \u00C1: 10.319987,
  \u00C0: 10.319987,
  \u00C4: 10.319987,
  \u00C2: 10.319987,
  \u00C3: 10.319987,
  \u00C5: 10.319987,
  B: 10,
  C: 9.663981,
  \u00C7: 9.663981,
  D: 10.191996,
  E: 9.408,
  \u00C9: 9.408,
  \u00C8: 9.408,
  \u00CB: 9.408,
  \u00CA: 9.408,
  F: 8.943991,
  G: 10.112,
  H: 11.055989,
  I: 6.175985,
  \u00CD: 6.175985,
  \u00CC: 6.175985,
  \u00CF: 6.175985,
  J: 7.391988,
  K: 10,
  L: 8.399984,
  " ": 4.095988,
  M: 13.503988,
  N: 11.199992,
  O: 10.447999,
  \u00D3: 10.447999,
  \u00D2: 10.447999,
  \u00D6: 10.447999,
  \u00D4: 10.447999,
  \u00D5: 10.447999,
  \u00D8: 10.447999,
  P: 9.77598,
  Q: 10.447999,
  R: 10.255981,
  S: 9.695984,
  \u0160: 9.695984,
  T: 8.767985,
  U: 10.607992,
  V: 10.207988,
  W: 15.519999,
  X: 10.671997,
  Y: 9.647989,
  \u00DD: 9.647989,
  \u0178: 9.647989,
  "!": 4.687988,
  '"': 6.319987,
  "'": 3.695984,
  "#": 10.175985,
  $: 9.359985,
  "%": 11.727987,
  "&": 10.063985,
  "(": 5.695984,
  ")": 5.503988,
  "*": 8.527995,
  "+": 9.199992,
  ",": 4.112,
  "-": 7.631999,
  ".": 4.112,
  "/": 5.823995,
  ":": 4.112,
  ";": 4.112,
  "<": 9.471985,
  "=": 9.199992,
  ">": 9.471985,
  "?": 8.879984,
  "@": 17.055989,
  "[": 5.839987,
  "\\": 5.823995,
  "]": 5.759989,
  "^": 9.199992,
  _: 9.279988,
  "`": 4.687988,
  "{": 6.255981,
  "|": 4.447999,
  "}": 6.255981,
  "~": 9.199992,
  "\u2022": 7.007996,
  "\u2026": 12.335988,
  "\xB0": 6.111979,
  \u00B5: 9.263997,
  "\xB2": 5.567993,
  "\xB3": 5.567993,
  "\xB1": 9.199992,
  "\u20AC": 9.359985,
  "\xA3": 9.359985,
  "\xA5": 9.199992
};
