// @dynatrace/strato-components/charts/core/components/text-ellipsis/truncate-string.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var truncate_string_exports = {};
__export(truncate_string_exports, {
  TEXT_HEIGHT: () => TEXT_HEIGHT,
  calculateWidth: () => calculateWidth,
  truncateString: () => truncateString
});
module.exports = __toCommonJS(truncate_string_exports);
var import_utils = require("../../../../typography/utils.js");
const TEXT_HEIGHT = 12;
const calculateWidth = (text, fontWidthMap) => {
  let totalWidth = 0;
  for (const char of text) {
    totalWidth += fontWidthMap[char] ?? fontWidthMap["@"];
  }
  return totalWidth;
};
function accumulateUntilCapacity(availableWidth, fontWidthMap, generator) {
  let keptText = "";
  let keptWidth = 0;
  const capacity = availableWidth - calculateWidth(import_utils.ELLIPSIS_CHARACTER, fontWidthMap);
  for (const ch of generator) {
    const chWidth = calculateWidth(ch, fontWidthMap);
    if (keptWidth + chWidth > capacity) {
      return keptText;
    }
    keptText += ch;
    keptWidth += chWidth;
  }
  return keptText;
}
const EPSILON_PX = 0.5;
const MIN_AVAILABLE_WIDTH = 19;
const buildTruncatedOutput = {
  start: (kept) => {
    let reversed = "";
    for (const ch of kept) {
      reversed = ch + reversed;
    }
    return `${import_utils.ELLIPSIS_CHARACTER}${reversed}`;
  },
  end: (kept) => `${kept}${import_utils.ELLIPSIS_CHARACTER}`,
  middle: (kept, priority) => {
    let head = "";
    let tail = "";
    if (priority === "start") {
      for (let i = 0; i < kept.length; i++) {
        if (i % 2 === 0) {
          head += kept[i];
        } else {
          tail = kept[i] + tail;
        }
      }
    } else {
      for (let i = 0; i < kept.length; i++) {
        if (i % 2 === 0) {
          tail = kept[i] + tail;
        } else {
          head += kept[i];
        }
      }
    }
    return `${head}${import_utils.ELLIPSIS_CHARACTER}${tail}`;
  }
};
const keepSequenceGenerators = {
  start: function* fromTail(text) {
    for (let i = text.length - 1; i >= 0; i--) {
      yield text[i];
    }
  },
  end: function* fromHead(text) {
    for (const ch of text) {
      yield ch;
    }
  },
  middle: function* fromBothEnds(text, priority) {
    let left = 0;
    let right = text.length - 1;
    while (left < right) {
      if (priority === "start") {
        yield text[left++];
        yield text[right--];
      } else {
        yield text[right--];
        yield text[left++];
      }
    }
    if (left === right) {
      yield text[left];
    }
  }
};
const truncateString = (text, availableWidth, fontWidthMap, truncationMode = "middle", priority = "end") => {
  const fullTextWidth = calculateWidth(text, fontWidthMap);
  const roundedAvailableWidth = Math.round(availableWidth) + EPSILON_PX;
  if (fullTextWidth <= roundedAvailableWidth) {
    return text;
  }
  if (MIN_AVAILABLE_WIDTH >= roundedAvailableWidth) {
    return import_utils.ELLIPSIS_CHARACTER;
  }
  const keptChars = accumulateUntilCapacity(
    availableWidth,
    fontWidthMap,
    truncationMode === "middle" ? keepSequenceGenerators.middle(text, priority) : keepSequenceGenerators[truncationMode](text)
  );
  return truncationMode === "middle" ? buildTruncatedOutput.middle(keptChars, priority) : buildTruncatedOutput[truncationMode](keptChars);
};
