// @dynatrace/strato-components/esm/charts/core/components/axes/components/utils/getCategoryBottomAxisProps.js
import {
  axisFontProps,
  TICK_LINE_HEIGHT,
  TICK_LINE_ICON_OFFSET
} from "../../constants.js";
const VISUAL_ADJUSTMENT = 2;
const getCategoryBottomAxisProps = (renderX, isTickRotated, maxSize, tickNameHasEmoji) => {
  const TICK_LINE_HEIGHT_VALIDATED = tickNameHasEmoji ? TICK_LINE_HEIGHT + TICK_LINE_ICON_OFFSET : TICK_LINE_HEIGHT;
  const x = isTickRotated ? -(maxSize + VISUAL_ADJUSTMENT * 2) : renderX - maxSize / VISUAL_ADJUSTMENT;
  const y = isTickRotated ? renderX - TICK_LINE_HEIGHT_VALIDATED / 2 : VISUAL_ADJUSTMENT;
  const foreignObjectInlineStyle = {
    height: TICK_LINE_HEIGHT_VALIDATED,
    width: maxSize,
    transform: isTickRotated ? "rotate(-90deg)" : "none"
  };
  const foreignObjectContentInlineStyle = {
    ...axisFontProps,
    textAlign: isTickRotated ? "right" : "center",
    height: TICK_LINE_HEIGHT_VALIDATED,
    width: maxSize,
    whiteSpace: "nowrap"
  };
  return { x, y, foreignObjectInlineStyle, foreignObjectContentInlineStyle };
};
export {
  getCategoryBottomAxisProps
};
