// @dynatrace-sdk/client-iam/esm/index.js
/**
 * @license
 * Copyright 2023-2026 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// packages/client/iam/src/lib/apis/users-and-groups-api.ts
import {
  httpClient as defaultHttpClient,
  isHttpClientAbortError,
  isHttpClientResponseError
} from "@dynatrace-sdk/http-client";
import {
  ApiClientError,
  ClientRequestError as ClientRequestError3,
  InvalidResponseError,
  isInvalidResponseError
} from "@dynatrace-sdk/shared-errors";

// packages/client/iam/src/lib/error-envelopes/api-gateway-errors-handler.ts
import { ApiGatewayError } from "@dynatrace-sdk/shared-errors";
async function apiGatewayErrorsHandler(response) {
  if (response.headers?.["dynatrace-response-source"] === "API Gateway") {
    const responseBody = await response.body("json");
    throw new ApiGatewayError(response, responseBody);
  }
}

// packages/client/iam/src/lib/error-envelopes/get-error-message.ts
function serializeData(data) {
  try {
    return JSON.stringify(data);
  } catch (e) {
    return String(data);
  }
}
function getMessagesFromErrorDetails(details) {
  const messages = [];
  Object.entries(details).forEach(([name, data]) => {
    if (!data)
      return;
    const serializedData = serializeData(data);
    switch (name) {
      case "missingScopes":
        messages.push(`Missing scopes: ${serializedData}`);
        break;
      default:
        messages.push(`${name}: ${serializedData}`);
    }
  }, []);
  return messages;
}
function getErrorMessage(errorBody, defaultMessage) {
  const error = errorBody;
  const msg = error?.error?.message || defaultMessage;
  const details = error?.error?.details || {};
  return [msg, ...getMessagesFromErrorDetails(details)].join(". ");
}

// packages/client/iam/src/lib/error-envelopes/public-exception-message-error.ts
import { ClientRequestError } from "@dynatrace-sdk/shared-errors";
var PublicExceptionMessageError = class extends ClientRequestError {
  isPublicExceptionMessageError = true;
};
function isPublicExceptionMessageError(e) {
  return e?.isApiClientError === true && e?.isClientRequestError === true && e?.isPublicExceptionMessageError === true && e instanceof Error;
}

// packages/client/iam/src/lib/error-envelopes/public-exception-throttling-message-error.ts
import { ClientRequestError as ClientRequestError2 } from "@dynatrace-sdk/shared-errors";
var PublicExceptionThrottlingMessageError = class extends ClientRequestError2 {
  isPublicExceptionThrottlingMessageError = true;
};
function isPublicExceptionThrottlingMessageError(e) {
  return e?.isApiClientError === true && e?.isClientRequestError === true && e?.isPublicExceptionThrottlingMessageError === true && e instanceof Error;
}

// packages/client/iam/src/lib/utils/transformations.ts
var defaultRuleConfig = {
  force: false,
  datetime: false
};
function mergeKeys(key1, key2) {
  if (!key1)
    return key2;
  if (!key2)
    return key1;
  return `${key1}.${key2}`;
}
function matchKey(key, rules, cache) {
  if (cache.has(key)) {
    return cache.get(key);
  }
  for (const rule of rules) {
    const pattern = Array.isArray(rule) ? rule[0] : rule;
    const config = Array.isArray(rule) ? rule[1] : defaultRuleConfig;
    if (key === pattern) {
      const result = [pattern, config];
      cache.set(key, result);
      return result;
    }
    let i = 0;
    let j = 0;
    const keyLen = key.length;
    const patternLen = pattern.length;
    let isMatch = true;
    while (i < keyLen && j < patternLen) {
      let keyPartEnd = key.indexOf(".", i);
      let patternPartEnd = pattern.indexOf(".", j);
      if (keyPartEnd === -1)
        keyPartEnd = keyLen;
      if (patternPartEnd === -1)
        patternPartEnd = patternLen;
      const keyPart = key.slice(i, keyPartEnd);
      const patternPart = pattern.slice(j, patternPartEnd);
      if (patternPart !== "*" && patternPart !== "**" && keyPart !== patternPart) {
        isMatch = false;
        break;
      }
      i = keyPartEnd + 1;
      j = patternPartEnd + 1;
    }
    if (isMatch) {
      if (j < patternLen) {
        const remaining = pattern.slice(j);
        if (remaining !== "**" && remaining !== "*") {
          isMatch = false;
        }
      }
      if (i < keyLen && pattern.slice(j - 1) !== "**") {
        isMatch = false;
      }
    }
    if (isMatch) {
      const result = [pattern, config];
      cache.set(key, result);
      return result;
    }
  }
  cache.set(key, null);
  return null;
}
var potentialMatchForRoot = (pattern, key) => pattern.startsWith(key) || pattern.startsWith("*") || pattern.startsWith("**");
var potentialMatch = (pattern, key) => pattern.startsWith(key) || pattern.includes("*") || pattern.includes("**");
function partialMatch(key, keyPatterns) {
  const keyPatternsLen = keyPatterns.length;
  if (!key.includes(".")) {
    for (let i = 0; i < keyPatternsLen; i++) {
      const pattern = keyPatterns[i];
      if (potentialMatchForRoot(pattern, key))
        return true;
    }
    return false;
  }
  let dotIndex = -1;
  let currentKeyEnd = 0;
  while ((dotIndex = key.indexOf(".", currentKeyEnd)) !== -1) {
    const currentKey = key.slice(0, dotIndex);
    const isRoot = !currentKey.includes(".");
    for (let i = 0; i < keyPatternsLen; i++) {
      const pattern = keyPatterns[i];
      if (isRoot) {
        if (potentialMatchForRoot(pattern, currentKey))
          return true;
      } else if (potentialMatch(pattern, currentKey))
        return true;
    }
    currentKeyEnd = dotIndex + 1;
  }
  for (let i = 0; i < keyPatternsLen; i++) {
    const pattern = keyPatterns[i];
    if (potentialMatch(pattern, key))
      return true;
  }
  return false;
}
var dateTimePattern = /^(\d{4}-\d{2}-\d{2})(T\d{2}:\d{2}:\d{2}(?:\.\d+)?(Z|[+-]\d{2}:\d{2})?)?$/i;
function isDateLike(value) {
  if (typeof value === "string") {
    return dateTimePattern.test(value);
  }
  return false;
}
function transformValue(value, rules, direction) {
  const { force, datetime } = rules;
  if (direction === "to") {
    return force || isDateLike(value) ? new Date(value) : value;
  }
  if (value instanceof Date) {
    const iso = value.toISOString();
    return datetime ? iso : iso.split("T")[0];
  }
  return value;
}
function walk(obj, direction, rules, rulesCache, keyPatterns, currentKey = "") {
  if (rules.length <= 0) {
    return obj;
  }
  if (!partialMatch(currentKey, keyPatterns)) {
    return obj;
  }
  if (Array.isArray(obj)) {
    const match = matchKey(currentKey, rules, rulesCache);
    obj.forEach((item, idx) => {
      if (typeof item === "object" && item !== null) {
        walk(item, direction, rules, rulesCache, keyPatterns, currentKey);
      } else if (match) {
        obj[idx] = transformValue(obj[idx], match[1], direction);
      }
    });
  }
  if (typeof obj === "object" && obj !== null && !Array.isArray(obj)) {
    for (const key of Object.keys(obj)) {
      const mergedKey = mergeKeys(currentKey, key);
      const match = matchKey(mergedKey, rules, rulesCache);
      if (match) {
        obj[key] = transformValue(obj[key], match[1], direction);
      }
      walk(obj[key], direction, rules, rulesCache, keyPatterns, mergedKey);
    }
  }
  return obj;
}
function flattenKeyPatterns(keyPatterns) {
  return keyPatterns.map((k) => Array.isArray(k) ? k[0] : k);
}
function transform(direction, object, rules) {
  const flatKeyPatterns = flattenKeyPatterns(rules);
  const rulesCache = /* @__PURE__ */ new Map();
  return walk(object, direction, rules, rulesCache, flatKeyPatterns);
}
function transformResponse(object, rules) {
  return transform("to", object, rules);
}

// packages/client/iam/src/lib/utils/url-helpers.ts
var encodeQueryParam = (key, value) => {
  const encodedKey = encodeURIComponent(key);
  return `${encodedKey}=${encodeURIComponent(typeof value === "number" ? value : String(value))}`;
};
var addExplodedArrayQueryParam = (query, key) => {
  const arrayValue = query[key];
  return arrayValue.map((value) => encodeQueryParam(key, value)).join("&");
};
var addNonExplodedArrayQueryParams = (query, key) => {
  const encodedKey = encodeURIComponent(key);
  const encodedParamsList = query[key].map((value) => encodeURIComponent(typeof value === "number" ? value : String(value))).join(",");
  return `${encodedKey}=${encodedParamsList}`;
};
var addQueryParam = (query, key) => encodeQueryParam(key, query[key]);
var arrayQueryParams = (query, key, explode) => {
  return explode ? addExplodedArrayQueryParam(query, key) : addNonExplodedArrayQueryParams(query, key);
};
var toQueryString = (rawQuery, flags = {}) => {
  const query = rawQuery || {};
  const keys = Object.keys(query).filter((key) => typeof query[key] !== "undefined");
  const queryString = keys.map(
    (key) => Array.isArray(query[key]) ? arrayQueryParams(query, key, flags.explode?.hasOwnProperty(key) ? flags.explode[key] : true) : addQueryParam(query, key)
  ).join("&");
  return queryString ? `?${queryString}` : "";
};

// packages/client/iam/src/lib/apis/users-and-groups-api.ts
var UsersAndGroupsClient = class {
  httpClient;
  shouldTransformDates = true;
  /**
   * @param {HttpClient} httpClientImplementation - You can provide custom http client as a parameter to constructor of a UsersAndGroupsClient. Custom http client should implement {HttpClient} interface.
   * @param {ApiClientOptions} [options] - Optional configuration for SDK Client.
   * @example
   * const usersAndGroupsClientNodeJs = new UsersAndGroupsClient(yourCustomImplementation);
   */
  constructor(httpClientImplementation, options) {
    this.httpClient = httpClientImplementation;
    this.shouldTransformDates = options?.transformDates ?? true;
  }
  /**
   * Get active user from organizational level
   *
   * **Required scope:** iam:users:read<br/><br/>Authorization is based on an assignment of the calling user to the account associated with the 'level-type' and 'level-id' For environment 'level-type' user has to be assigned to account to which the environment is assigned, user does not need any permissions in such environment
   *
   * @returns {Promise<RestUserPublic>} User fetched successfully
   *
   * @example <caption>Code example</caption>
   * import { usersAndGroupsClient } from "@dynatrace-sdk/client-iam";
   *
   * const data =
   *   await usersAndGroupsClient.getActiveUserFromOrganizationalLevel(
   *     { levelType: "...", levelId: "...", uuid: "..." },
   *   );
   *
   * @throws {PublicExceptionMessageError} Authorization header is missing or is incorrect. | Execution user is not allowed to perform the action. | Calling user is not assigned to the requested account or requested environment is not assigned to account that user is assigned to or the 'level-id' does not exist
   * @throws {PublicExceptionThrottlingMessageError} Too many requests have been sent - requests are throttled
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async getActiveUserFromOrganizationalLevel(config) {
    try {
      const response = await this.httpClient.send({
        url: `/platform/iam/v1/organizational-levels/${config.levelType}/${config.levelId}/users/${config.uuid}`,
        method: "GET",
        headers: {
          Accept: "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
      } catch (err) {
        throw new InvalidResponseError(
          `UsersAndGroupsClient.getActiveUserFromOrganizationalLevel:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if (isHttpClientAbortError(e) || isInvalidResponseError(e)) {
        throw e;
      }
      if (!isHttpClientResponseError(e)) {
        throw new ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      switch (response.status) {
        case 401: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new PublicExceptionMessageError(
              `401`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Authorization header is missing or is incorrect."),
              e
            );
          } catch (err) {
            if (err instanceof PublicExceptionMessageError) {
              throw err;
            }
            throw new InvalidResponseError(
              `UsersAndGroupsClient.getActiveUserFromOrganizationalLevel:401`,
              err,
              responseValue,
              "PublicExceptionMessage",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new PublicExceptionMessageError(
              `403`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Execution user is not allowed to perform the action."),
              e
            );
          } catch (err) {
            if (err instanceof PublicExceptionMessageError) {
              throw err;
            }
            throw new InvalidResponseError(
              `UsersAndGroupsClient.getActiveUserFromOrganizationalLevel:403`,
              err,
              responseValue,
              "PublicExceptionMessage",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new PublicExceptionMessageError(
              `404`,
              response,
              errorBody,
              getErrorMessage(
                errorBody,
                "Calling user is not assigned to the requested account or requested environment is not assigned to account that user is assigned to or the 'level-id' does not exist"
              ),
              e
            );
          } catch (err) {
            if (err instanceof PublicExceptionMessageError) {
              throw err;
            }
            throw new InvalidResponseError(
              `UsersAndGroupsClient.getActiveUserFromOrganizationalLevel:404`,
              err,
              responseValue,
              "PublicExceptionMessage",
              void 0,
              response
            );
          }
        }
        case 429: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new PublicExceptionThrottlingMessageError(
              `429`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Too many requests have been sent - requests are throttled"),
              e
            );
          } catch (err) {
            if (err instanceof PublicExceptionThrottlingMessageError) {
              throw err;
            }
            throw new InvalidResponseError(
              `UsersAndGroupsClient.getActiveUserFromOrganizationalLevel:429`,
              err,
              responseValue,
              "PublicExceptionThrottlingMessage",
              void 0,
              response
            );
          }
        }
        default: {
          try {
            const responseValue = await response.body("json");
            throw new ClientRequestError3(
              `${response.status}`,
              response,
              responseValue,
              getErrorMessage(
                responseValue,
                `Unexpected api response: code=${response.status} body="${responseValue}"`
              ),
              e
            );
          } catch (error) {
            if (error instanceof ClientRequestError3) {
              throw error;
            }
            const responseValue = await response.body("text").catch(() => "");
            throw new InvalidResponseError(
              `${response.status}`,
              error,
              responseValue,
              "json",
              "Unexpected api response",
              response
            );
          }
        }
      }
    }
  }
  /**
   * Get active users from organizational level
   *
   * **Required scope:** iam:users:read<br/><br/>Providing value for at least one of the parameters: 'partialString' or 'uuid' is required.
   * If both query params 'partialString' and 'uuid' are provided then the result returns users meeting any of the filtering criteria.
   * Returned users are first ordered by name and then by surname.
   * Authorization is based on an assignment of the calling user to the account associated with the 'level-type' and 'level-id' For environment 'level-type' user has to be assigned to account to which the environment is assigned, user does not need any permissions in such environment
   *
   * @returns {Promise<RestUserPublicListResponse>} Users from the organizational level fetched successfully
   *
   * @example <caption>Code example</caption>
   * import { usersAndGroupsClient } from "@dynatrace-sdk/client-iam";
   *
   * const data =
   *   await usersAndGroupsClient.getActiveUsersForOrganizationalLevel(
   *     { levelType: "...", levelId: "..." },
   *   );
   *
   * @throws {PublicExceptionMessageError} Authorization header is missing or is incorrect. | Execution user is not allowed to perform the action. | Calling user is not assigned to the requested account or requested environment is not assigned to account that user is assigned to or the 'level-id' does not exist
   * @throws {PublicExceptionThrottlingMessageError} Too many requests have been sent - requests are throttled
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async getActiveUsersForOrganizationalLevel(config) {
    if (!config) {
      throw new ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const query = toQueryString({
      partialString: config.partialString,
      uuid: config.uuid,
      "page-key": config.pageKey,
      page: config.page,
      pageSize: config.pageSize
    });
    try {
      const response = await this.httpClient.send({
        url: `/platform/iam/v1/organizational-levels/${config.levelType}/${config.levelId}/users${query}`,
        method: "GET",
        headers: {
          Accept: "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
      } catch (err) {
        throw new InvalidResponseError(
          `UsersAndGroupsClient.getActiveUsersForOrganizationalLevel:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if (isHttpClientAbortError(e) || isInvalidResponseError(e)) {
        throw e;
      }
      if (!isHttpClientResponseError(e)) {
        throw new ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      switch (response.status) {
        case 401: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new PublicExceptionMessageError(
              `401`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Authorization header is missing or is incorrect."),
              e
            );
          } catch (err) {
            if (err instanceof PublicExceptionMessageError) {
              throw err;
            }
            throw new InvalidResponseError(
              `UsersAndGroupsClient.getActiveUsersForOrganizationalLevel:401`,
              err,
              responseValue,
              "PublicExceptionMessage",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new PublicExceptionMessageError(
              `403`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Execution user is not allowed to perform the action."),
              e
            );
          } catch (err) {
            if (err instanceof PublicExceptionMessageError) {
              throw err;
            }
            throw new InvalidResponseError(
              `UsersAndGroupsClient.getActiveUsersForOrganizationalLevel:403`,
              err,
              responseValue,
              "PublicExceptionMessage",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new PublicExceptionMessageError(
              `404`,
              response,
              errorBody,
              getErrorMessage(
                errorBody,
                "Calling user is not assigned to the requested account or requested environment is not assigned to account that user is assigned to or the 'level-id' does not exist"
              ),
              e
            );
          } catch (err) {
            if (err instanceof PublicExceptionMessageError) {
              throw err;
            }
            throw new InvalidResponseError(
              `UsersAndGroupsClient.getActiveUsersForOrganizationalLevel:404`,
              err,
              responseValue,
              "PublicExceptionMessage",
              void 0,
              response
            );
          }
        }
        case 429: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new PublicExceptionThrottlingMessageError(
              `429`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Too many requests have been sent - requests are throttled"),
              e
            );
          } catch (err) {
            if (err instanceof PublicExceptionThrottlingMessageError) {
              throw err;
            }
            throw new InvalidResponseError(
              `UsersAndGroupsClient.getActiveUsersForOrganizationalLevel:429`,
              err,
              responseValue,
              "PublicExceptionThrottlingMessage",
              void 0,
              response
            );
          }
        }
        default: {
          try {
            const responseValue = await response.body("json");
            throw new ClientRequestError3(
              `${response.status}`,
              response,
              responseValue,
              getErrorMessage(
                responseValue,
                `Unexpected api response: code=${response.status} body="${responseValue}"`
              ),
              e
            );
          } catch (error) {
            if (error instanceof ClientRequestError3) {
              throw error;
            }
            const responseValue = await response.body("text").catch(() => "");
            throw new InvalidResponseError(
              `${response.status}`,
              error,
              responseValue,
              "json",
              "Unexpected api response",
              response
            );
          }
        }
      }
    }
  }
  /**
   * Get active users from organizational level
   *
   * **Required scope:** iam:users:read<br/><br/>It is required to provide either request body containing list of user UUIDs or partialString.
   * If both are provided then the result returns users meeting any of the provided criteria.
   * Returned users are first ordered by name and then by surname.
   * Authorization is based on an assignment of the calling user to the account associated with the 'level-type' and 'level-id' For environment 'level-type' user has to be assigned to account to which the environment is assigned, user does not need any permissions in such environment
   *
   * @returns {Promise<RestUserPublicListResponse>} Users from the organizational level fetched successfully
   *
   * @example <caption>Code example</caption>
   * import { usersAndGroupsClient } from "@dynatrace-sdk/client-iam";
   *
   * const data =
   *   await usersAndGroupsClient.getActiveUsersForOrganizationalLevelPost(
   *     { levelType: "...", levelId: "...", body: ["..."] },
   *   );
   *
   * @throws {PublicExceptionMessageError} Authorization header is missing or is incorrect. | Execution user is not allowed to perform the action. | Calling user is not assigned to the requested account or requested environment is not assigned to account that user is assigned to or the 'level-id' does not exist
   * @throws {PublicExceptionThrottlingMessageError} Too many requests have been sent - requests are throttled
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async getActiveUsersForOrganizationalLevelPost(config) {
    if (!config) {
      throw new ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const query = toQueryString({
      partialString: config.partialString,
      "page-key": config.pageKey,
      page: config.page,
      pageSize: config.pageSize
    });
    try {
      const response = await this.httpClient.send({
        url: `/platform/iam/v1/organizational-levels/${config.levelType}/${config.levelId}/users${query}`,
        method: "POST",
        requestBodyType: "json",
        body: config.body,
        headers: {
          "Content-Type": "application/json",
          Accept: "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
      } catch (err) {
        throw new InvalidResponseError(
          `UsersAndGroupsClient.getActiveUsersForOrganizationalLevelPOST:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if (isHttpClientAbortError(e) || isInvalidResponseError(e)) {
        throw e;
      }
      if (!isHttpClientResponseError(e)) {
        throw new ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      switch (response.status) {
        case 401: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new PublicExceptionMessageError(
              `401`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Authorization header is missing or is incorrect."),
              e
            );
          } catch (err) {
            if (err instanceof PublicExceptionMessageError) {
              throw err;
            }
            throw new InvalidResponseError(
              `UsersAndGroupsClient.getActiveUsersForOrganizationalLevelPOST:401`,
              err,
              responseValue,
              "PublicExceptionMessage",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new PublicExceptionMessageError(
              `403`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Execution user is not allowed to perform the action."),
              e
            );
          } catch (err) {
            if (err instanceof PublicExceptionMessageError) {
              throw err;
            }
            throw new InvalidResponseError(
              `UsersAndGroupsClient.getActiveUsersForOrganizationalLevelPOST:403`,
              err,
              responseValue,
              "PublicExceptionMessage",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new PublicExceptionMessageError(
              `404`,
              response,
              errorBody,
              getErrorMessage(
                errorBody,
                "Calling user is not assigned to the requested account or requested environment is not assigned to account that user is assigned to or the 'level-id' does not exist"
              ),
              e
            );
          } catch (err) {
            if (err instanceof PublicExceptionMessageError) {
              throw err;
            }
            throw new InvalidResponseError(
              `UsersAndGroupsClient.getActiveUsersForOrganizationalLevelPOST:404`,
              err,
              responseValue,
              "PublicExceptionMessage",
              void 0,
              response
            );
          }
        }
        case 429: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new PublicExceptionThrottlingMessageError(
              `429`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Too many requests have been sent - requests are throttled"),
              e
            );
          } catch (err) {
            if (err instanceof PublicExceptionThrottlingMessageError) {
              throw err;
            }
            throw new InvalidResponseError(
              `UsersAndGroupsClient.getActiveUsersForOrganizationalLevelPOST:429`,
              err,
              responseValue,
              "PublicExceptionThrottlingMessage",
              void 0,
              response
            );
          }
        }
        default: {
          try {
            const responseValue = await response.body("json");
            throw new ClientRequestError3(
              `${response.status}`,
              response,
              responseValue,
              getErrorMessage(
                responseValue,
                `Unexpected api response: code=${response.status} body="${responseValue}"`
              ),
              e
            );
          } catch (error) {
            if (error instanceof ClientRequestError3) {
              throw error;
            }
            const responseValue = await response.body("text").catch(() => "");
            throw new InvalidResponseError(
              `${response.status}`,
              error,
              responseValue,
              "json",
              "Unexpected api response",
              response
            );
          }
        }
      }
    }
  }
  /**
   * Get active service users from organizational level which are usable by the execution user
   *
   * **Required scope:** iam:service-users:use<br/><br/>Environment-level queries are executed in account context because there is no direct link between environments and service users.
   *
   * @returns {Promise<SearchResult>} Active service users from the organizational level filtered by user permissions fetched successfully
   *
   * @example <caption>Code example</caption>
   * import { usersAndGroupsClient } from "@dynatrace-sdk/client-iam";
   *
   * const data =
   *   await usersAndGroupsClient.getAvailableServiceUsers({
   *     levelType: "...",
   *     levelId: "...",
   *   });
   *
   * @throws {PublicExceptionMessageError} Authorization header is missing or is incorrect. | Execution user is not allowed to perform the action. | The 'level-id' param does not exist
   * @throws {PublicExceptionThrottlingMessageError} Too many requests have been sent - requests are throttled
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async getAvailableServiceUsers(config) {
    if (!config) {
      throw new ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const query = toQueryString({ "page-key": config.pageKey, page: config.page, pageSize: config.pageSize });
    try {
      const response = await this.httpClient.send({
        url: `/platform/iam/v1/organizational-levels/${config.levelType}/${config.levelId}/service-users${query}`,
        method: "GET",
        headers: {
          Accept: "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? transformResponse(responseValue, [["results.createdAt", { datetime: true }]]) : responseValue;
      } catch (err) {
        throw new InvalidResponseError(
          `UsersAndGroupsClient.getAvailableServiceUsers:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if (isHttpClientAbortError(e) || isInvalidResponseError(e)) {
        throw e;
      }
      if (!isHttpClientResponseError(e)) {
        throw new ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      switch (response.status) {
        case 401: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new PublicExceptionMessageError(
              `401`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Authorization header is missing or is incorrect."),
              e
            );
          } catch (err) {
            if (err instanceof PublicExceptionMessageError) {
              throw err;
            }
            throw new InvalidResponseError(
              `UsersAndGroupsClient.getAvailableServiceUsers:401`,
              err,
              responseValue,
              "PublicExceptionMessage",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new PublicExceptionMessageError(
              `403`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Execution user is not allowed to perform the action."),
              e
            );
          } catch (err) {
            if (err instanceof PublicExceptionMessageError) {
              throw err;
            }
            throw new InvalidResponseError(
              `UsersAndGroupsClient.getAvailableServiceUsers:403`,
              err,
              responseValue,
              "PublicExceptionMessage",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new PublicExceptionMessageError(
              `404`,
              response,
              errorBody,
              getErrorMessage(errorBody, "The 'level-id' param does not exist"),
              e
            );
          } catch (err) {
            if (err instanceof PublicExceptionMessageError) {
              throw err;
            }
            throw new InvalidResponseError(
              `UsersAndGroupsClient.getAvailableServiceUsers:404`,
              err,
              responseValue,
              "PublicExceptionMessage",
              void 0,
              response
            );
          }
        }
        case 429: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new PublicExceptionThrottlingMessageError(
              `429`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Too many requests have been sent - requests are throttled"),
              e
            );
          } catch (err) {
            if (err instanceof PublicExceptionThrottlingMessageError) {
              throw err;
            }
            throw new InvalidResponseError(
              `UsersAndGroupsClient.getAvailableServiceUsers:429`,
              err,
              responseValue,
              "PublicExceptionThrottlingMessage",
              void 0,
              response
            );
          }
        }
        default: {
          try {
            const responseValue = await response.body("json");
            throw new ClientRequestError3(
              `${response.status}`,
              response,
              responseValue,
              getErrorMessage(
                responseValue,
                `Unexpected api response: code=${response.status} body="${responseValue}"`
              ),
              e
            );
          } catch (error) {
            if (error instanceof ClientRequestError3) {
              throw error;
            }
            const responseValue = await response.body("text").catch(() => "");
            throw new InvalidResponseError(
              `${response.status}`,
              error,
              responseValue,
              "json",
              "Unexpected api response",
              response
            );
          }
        }
      }
    }
  }
  /**
   * Get visible groups from organizational level
   *
   * **Required scope:** iam:groups:read<br/><br/>Providing value for at least one of the parameters: 'partialGroupName' or 'uuid' is required.
   * If both query params 'partialGroupName' and 'uuid' are provided then the result will be all of the groups meeting any of the filtering criteria.
   * Authorization is based on an assignment of the calling user to the account associated with the 'level-type' and 'level-id' For environment 'level-type' user has to be assigned to account to which the environment is assigned, user does not need any permissions in such environment
   *
   * @returns {Promise<RestGroupPublicListResponse>} Groups from the level fetched successfully
   *
   * @example <caption>Code example</caption>
   * import { usersAndGroupsClient } from "@dynatrace-sdk/client-iam";
   *
   * const data =
   *   await usersAndGroupsClient.getVisibleGroupsForAccount({
   *     levelType: "...",
   *     levelId: "...",
   *   });
   *
   * @throws {PublicExceptionMessageError} Missing both 'partialGroupName' and 'uuid' filter values, or the values don't meet the requirements. | Authorization header is missing or is incorrect. | Execution user is not allowed to perform the action. | Calling user is not assigned to the requested account or requested environment is not assigned to account that user is assigned to or the 'level-id' does not exist
   * @throws {PublicExceptionThrottlingMessageError} Too many requests have been sent - requests are throttled
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async getVisibleGroupsForAccount(config) {
    if (!config) {
      throw new ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const query = toQueryString({
      partialGroupName: config.partialGroupName,
      uuid: config.uuid,
      "page-key": config.pageKey,
      page: config.page,
      pageSize: config.pageSize
    });
    try {
      const response = await this.httpClient.send({
        url: `/platform/iam/v1/organizational-levels/${config.levelType}/${config.levelId}/groups${query}`,
        method: "GET",
        headers: {
          Accept: "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
      } catch (err) {
        throw new InvalidResponseError(
          `UsersAndGroupsClient.getVisibleGroupsForAccount:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if (isHttpClientAbortError(e) || isInvalidResponseError(e)) {
        throw e;
      }
      if (!isHttpClientResponseError(e)) {
        throw new ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new PublicExceptionMessageError(
              `400`,
              response,
              errorBody,
              getErrorMessage(
                errorBody,
                "Missing both 'partialGroupName' and 'uuid' filter values, or the values don't meet the requirements."
              ),
              e
            );
          } catch (err) {
            if (err instanceof PublicExceptionMessageError) {
              throw err;
            }
            throw new InvalidResponseError(
              `UsersAndGroupsClient.getVisibleGroupsForAccount:400`,
              err,
              responseValue,
              "PublicExceptionMessage",
              void 0,
              response
            );
          }
        }
        case 401: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new PublicExceptionMessageError(
              `401`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Authorization header is missing or is incorrect."),
              e
            );
          } catch (err) {
            if (err instanceof PublicExceptionMessageError) {
              throw err;
            }
            throw new InvalidResponseError(
              `UsersAndGroupsClient.getVisibleGroupsForAccount:401`,
              err,
              responseValue,
              "PublicExceptionMessage",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new PublicExceptionMessageError(
              `403`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Execution user is not allowed to perform the action."),
              e
            );
          } catch (err) {
            if (err instanceof PublicExceptionMessageError) {
              throw err;
            }
            throw new InvalidResponseError(
              `UsersAndGroupsClient.getVisibleGroupsForAccount:403`,
              err,
              responseValue,
              "PublicExceptionMessage",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new PublicExceptionMessageError(
              `404`,
              response,
              errorBody,
              getErrorMessage(
                errorBody,
                "Calling user is not assigned to the requested account or requested environment is not assigned to account that user is assigned to or the 'level-id' does not exist"
              ),
              e
            );
          } catch (err) {
            if (err instanceof PublicExceptionMessageError) {
              throw err;
            }
            throw new InvalidResponseError(
              `UsersAndGroupsClient.getVisibleGroupsForAccount:404`,
              err,
              responseValue,
              "PublicExceptionMessage",
              void 0,
              response
            );
          }
        }
        case 429: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new PublicExceptionThrottlingMessageError(
              `429`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Too many requests have been sent - requests are throttled"),
              e
            );
          } catch (err) {
            if (err instanceof PublicExceptionThrottlingMessageError) {
              throw err;
            }
            throw new InvalidResponseError(
              `UsersAndGroupsClient.getVisibleGroupsForAccount:429`,
              err,
              responseValue,
              "PublicExceptionThrottlingMessage",
              void 0,
              response
            );
          }
        }
        default: {
          try {
            const responseValue = await response.body("json");
            throw new ClientRequestError3(
              `${response.status}`,
              response,
              responseValue,
              getErrorMessage(
                responseValue,
                `Unexpected api response: code=${response.status} body="${responseValue}"`
              ),
              e
            );
          } catch (error) {
            if (error instanceof ClientRequestError3) {
              throw error;
            }
            const responseValue = await response.body("text").catch(() => "");
            throw new InvalidResponseError(
              `${response.status}`,
              error,
              responseValue,
              "json",
              "Unexpected api response",
              response
            );
          }
        }
      }
    }
  }
  /**
   * Get visible groups from organizational level
   *
   * **Required scope:** iam:groups:read<br/><br/>It is required to provide at either request body containing list of group uuids or partialGroupName.
   * If both are provided then the result returns groups meeting any of the provided criteria.
   * Authorization is based on an assignment of the calling user to the account associated with the 'level-type' and 'level-id' For environment 'level-type' user has to be assigned to account to which the environment is assigned, user does not need any permissions in such environment
   *
   * @returns {Promise<RestGroupPublicListResponse>} Groups from the level fetched successfully
   *
   * @example <caption>Code example</caption>
   * import { usersAndGroupsClient } from "@dynatrace-sdk/client-iam";
   *
   * const data =
   *   await usersAndGroupsClient.getVisibleGroupsForAccountPost(
   *     { levelType: "...", levelId: "...", body: ["..."] },
   *   );
   *
   * @throws {PublicExceptionMessageError} Missing both 'partialGroupName' and 'uuid' filter values, or the values don't meet requirements. | Authorization header is missing or is incorrect. | Execution user is not allowed to perform the action. | Calling user is not assigned to the requested account or requested environment is not assigned to account that user is assigned to or the 'level-id' does not exist
   * @throws {PublicExceptionThrottlingMessageError} Too many requests have been sent - requests are throttled
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async getVisibleGroupsForAccountPost(config) {
    if (!config) {
      throw new ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const query = toQueryString({
      partialGroupName: config.partialGroupName,
      "page-key": config.pageKey,
      page: config.page,
      pageSize: config.pageSize
    });
    try {
      const response = await this.httpClient.send({
        url: `/platform/iam/v1/organizational-levels/${config.levelType}/${config.levelId}/groups${query}`,
        method: "POST",
        requestBodyType: "json",
        body: config.body,
        headers: {
          "Content-Type": "application/json",
          Accept: "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
      } catch (err) {
        throw new InvalidResponseError(
          `UsersAndGroupsClient.getVisibleGroupsForAccountPOST:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if (isHttpClientAbortError(e) || isInvalidResponseError(e)) {
        throw e;
      }
      if (!isHttpClientResponseError(e)) {
        throw new ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new PublicExceptionMessageError(
              `400`,
              response,
              errorBody,
              getErrorMessage(
                errorBody,
                "Missing both 'partialGroupName' and 'uuid' filter values, or the values don't meet requirements."
              ),
              e
            );
          } catch (err) {
            if (err instanceof PublicExceptionMessageError) {
              throw err;
            }
            throw new InvalidResponseError(
              `UsersAndGroupsClient.getVisibleGroupsForAccountPOST:400`,
              err,
              responseValue,
              "PublicExceptionMessage",
              void 0,
              response
            );
          }
        }
        case 401: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new PublicExceptionMessageError(
              `401`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Authorization header is missing or is incorrect."),
              e
            );
          } catch (err) {
            if (err instanceof PublicExceptionMessageError) {
              throw err;
            }
            throw new InvalidResponseError(
              `UsersAndGroupsClient.getVisibleGroupsForAccountPOST:401`,
              err,
              responseValue,
              "PublicExceptionMessage",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new PublicExceptionMessageError(
              `403`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Execution user is not allowed to perform the action."),
              e
            );
          } catch (err) {
            if (err instanceof PublicExceptionMessageError) {
              throw err;
            }
            throw new InvalidResponseError(
              `UsersAndGroupsClient.getVisibleGroupsForAccountPOST:403`,
              err,
              responseValue,
              "PublicExceptionMessage",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new PublicExceptionMessageError(
              `404`,
              response,
              errorBody,
              getErrorMessage(
                errorBody,
                "Calling user is not assigned to the requested account or requested environment is not assigned to account that user is assigned to or the 'level-id' does not exist"
              ),
              e
            );
          } catch (err) {
            if (err instanceof PublicExceptionMessageError) {
              throw err;
            }
            throw new InvalidResponseError(
              `UsersAndGroupsClient.getVisibleGroupsForAccountPOST:404`,
              err,
              responseValue,
              "PublicExceptionMessage",
              void 0,
              response
            );
          }
        }
        case 429: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new PublicExceptionThrottlingMessageError(
              `429`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Too many requests have been sent - requests are throttled"),
              e
            );
          } catch (err) {
            if (err instanceof PublicExceptionThrottlingMessageError) {
              throw err;
            }
            throw new InvalidResponseError(
              `UsersAndGroupsClient.getVisibleGroupsForAccountPOST:429`,
              err,
              responseValue,
              "PublicExceptionThrottlingMessage",
              void 0,
              response
            );
          }
        }
        default: {
          try {
            const responseValue = await response.body("json");
            throw new ClientRequestError3(
              `${response.status}`,
              response,
              responseValue,
              getErrorMessage(
                responseValue,
                `Unexpected api response: code=${response.status} body="${responseValue}"`
              ),
              e
            );
          } catch (error) {
            if (error instanceof ClientRequestError3) {
              throw error;
            }
            const responseValue = await response.body("text").catch(() => "");
            throw new InvalidResponseError(
              `${response.status}`,
              error,
              responseValue,
              "json",
              "Unexpected api response",
              response
            );
          }
        }
      }
    }
  }
};
var usersAndGroupsClient = /* @__PURE__ */ new UsersAndGroupsClient(defaultHttpClient);

// packages/client/iam/src/lib/error-envelopes/index.ts
import {
  ApiClientError as ApiClientError2,
  ApiGatewayError as ApiGatewayError2,
  ClientRequestError as ClientRequestError4,
  InvalidResponseError as InvalidResponseError2,
  isApiClientError,
  isApiGatewayError,
  isClientRequestError,
  isInvalidResponseError as isInvalidResponseError2
} from "@dynatrace-sdk/shared-errors";

// packages/client/iam/src/lib/models/rest-group-public-type.ts
var RestGroupPublicType = /* @__PURE__ */ ((RestGroupPublicType2) => {
  RestGroupPublicType2["Local"] = "LOCAL";
  RestGroupPublicType2["AllUsers"] = "ALL_USERS";
  RestGroupPublicType2["Scim"] = "SCIM";
  RestGroupPublicType2["Saml"] = "SAML";
  return RestGroupPublicType2;
})(RestGroupPublicType || {});
export {
  ApiClientError2 as ApiClientError,
  ApiGatewayError2 as ApiGatewayError,
  ClientRequestError4 as ClientRequestError,
  InvalidResponseError2 as InvalidResponseError,
  PublicExceptionMessageError,
  PublicExceptionThrottlingMessageError,
  RestGroupPublicType,
  UsersAndGroupsClient,
  isApiClientError,
  isApiGatewayError,
  isClientRequestError,
  isInvalidResponseError2 as isInvalidResponseError,
  isPublicExceptionMessageError,
  isPublicExceptionThrottlingMessageError,
  usersAndGroupsClient
};
