// @dynatrace/strato-components/content/accordion/Accordion.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Accordion_exports = {};
__export(Accordion_exports, {
  Accordion: () => Accordion
});
module.exports = __toCommonJS(Accordion_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_accordion_state = require("./accordion-state.js");
var import_Accordion_css = require("./Accordion.sty.js");
var import_AccordionSection = require("./AccordionSection.js");
var import_AccordionSectionContent = require("./AccordionSectionContent.js");
var import_AccordionSectionLabel = require("./AccordionSectionLabel.js");
var import_AccordionContext = require("./contexts/AccordionContext.js");
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useSafeForwardProps = require("../../core/hooks/useSafeForwardProps.js");
var import_Divider = require("../../layouts/divider/Divider.js");
var import_Flex = require("../../layouts/flex/Flex.js");
const COMPONENT_NAME = "Accordion";
function RootAccordion(props, forwardedRef) {
  const {
    expanded,
    defaultExpanded,
    onExpandChange,
    multiple = false,
    triggerPosition = "end",
    interactive = true,
    color = "neutral",
    showDividers = true,
    size = "default",
    children,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const ariaLabelingProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(restProps);
  const forwardProps = (0, import_useSafeForwardProps.useSafeForwardProps)(restProps, ariaLabelingProps);
  const sectionCount = import_react.Children.count(children);
  const accordionSections = import_react.Children.map(children, (child, index) => /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    child,
    showDividers && sectionCount > 1 && index < sectionCount - 1 && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Divider.Divider, { className: import_Accordion_css.accordionDividerMarginCSS[size], color })
  ] }));
  const [state, dispatch] = (0, import_accordion_state.useAccordionState)({
    expanded,
    defaultExpanded,
    multiple,
    onExpandChange
  });
  const lastMultiple = (0, import_react.useRef)(multiple);
  if (multiple !== lastMultiple.current) {
    dispatch({ type: import_accordion_state.AccordionActions.SET_MULTIPLE, payload: multiple });
    lastMultiple.current = multiple;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      ref: forwardedRef,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...ariaLabelingProps,
      ...behavioralTrackingProps,
      ...forwardProps,
      "data-dt-component": COMPONENT_NAME,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_AccordionContext.AccordionContextProvider,
        {
          value: {
            triggerPosition,
            size,
            interactive,
            state,
            color,
            dispatch
          },
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_Flex.Flex,
            {
              flexDirection: "column",
              gap: showDividers ? 0 : size === "default" ? 8 : 4,
              children: accordionSections
            }
          )
        }
      )
    }
  );
}
RootAccordion.displayName = COMPONENT_NAME;
const AccordionForwardRef = (0, import_react.forwardRef)(RootAccordion);
const Accordion = Object.assign(AccordionForwardRef, {
  Section: import_AccordionSection.AccordionSection,
  SectionLabel: import_AccordionSectionLabel.AccordionSectionLabel,
  SectionContent: import_AccordionSectionContent.AccordionSectionContent
});
