// @dynatrace/strato-components/content/accordion/accordion-state.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var accordion_state_exports = {};
__export(accordion_state_exports, {
  AccordionActions: () => AccordionActions,
  useAccordionState: () => useAccordionState
});
module.exports = __toCommonJS(accordion_state_exports);
var import_react = require("react");
var AccordionActions = /* @__PURE__ */ ((AccordionActions2) => {
  AccordionActions2["TOGGLE_SECTION"] = "toggle-section";
  AccordionActions2["SET_MULTIPLE"] = "set-multiple";
  return AccordionActions2;
})(AccordionActions || {});
function convertPropToState(expanded) {
  const arrayExpand = Array.isArray(expanded) ? expanded : [expanded];
  const filteredExpand = arrayExpand.filter(
    (item) => item !== "" && item !== void 0
  );
  return {
    expanded: /* @__PURE__ */ new Set([...filteredExpand])
  };
}
function isEqual(setA, setB) {
  if (setA.size !== setB.size) {
    return false;
  }
  return [...setA].every((value) => setB.has(value));
}
function useAccordionState({
  expanded: expandedProp,
  defaultExpanded: defaultExpandedProp,
  multiple: multipleProp,
  onExpandChange: onExpandChangeProp
}) {
  const isControlled = expandedProp !== void 0;
  function emitOnChange(expandedState) {
    const expandedSection = Array.from(expandedState);
    if (multipleProp === false) {
      onExpandChangeProp?.(
        // Falls back to an empty string to not switch between controlled and uncontrolled
        expandedSection[0] ?? ""
      );
    } else {
      onExpandChangeProp?.(expandedSection);
    }
  }
  function stateReducer(state2, action) {
    const newExpanded = new Set(state2.expanded);
    if (action.type === "toggle-section" /* TOGGLE_SECTION */) {
      if (multipleProp === false && !newExpanded.has(action.payload)) {
        newExpanded.clear();
      }
      if (newExpanded.has(action.payload)) {
        newExpanded.delete(action.payload);
      } else {
        newExpanded.add(action.payload);
      }
      if (!isEqual(state2.expanded, newExpanded)) {
        emitOnChange(newExpanded);
      }
      if (!isControlled) {
        return {
          expanded: newExpanded
        };
      }
    }
    if (action.type === "set-multiple" /* SET_MULTIPLE */) {
      if (action.payload === false && state2.expanded.size > 1) {
        const firstExpanded = Array.from(state2.expanded)[0];
        newExpanded.clear();
        newExpanded.add(firstExpanded);
      }
      if (!isEqual(state2.expanded, newExpanded)) {
        emitOnChange(newExpanded);
      }
      if (!isControlled) {
        return {
          expanded: newExpanded
        };
      }
    }
    return state2;
  }
  const initExpanded = multipleProp ? [] : void 0;
  const defaultExpanded = defaultExpandedProp ?? initExpanded;
  const [uncontrolledState, uncontrolledDispatch] = (0, import_react.useReducer)(
    stateReducer,
    defaultExpanded,
    convertPropToState
  );
  const state = isControlled ? convertPropToState(expandedProp) : uncontrolledState;
  function controlledDispatch(action) {
    stateReducer(state, action);
  }
  return [state, isControlled ? controlledDispatch : uncontrolledDispatch];
}
