// @dynatrace/strato-components/content/accordion/AccordionSection.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var AccordionSection_exports = {};
__export(AccordionSection_exports, {
  AccordionSection: () => AccordionSection
});
module.exports = __toCommonJS(AccordionSection_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_use_resize_observer = __toESM(require("use-resize-observer"));
var import_strato_icons = require("@dynatrace/strato-icons");
var import_accordion_state = require("./accordion-state.js");
var import_Accordion_css = require("./Accordion.sty.js");
var import_AccordionSectionContent = require("./AccordionSectionContent.js");
var import_AccordionSectionLabel = require("./AccordionSectionLabel.js");
var import_AccordionContext = require("./contexts/AccordionContext.js");
var import_AccordionSectionContext = require("./contexts/AccordionSectionContext.js");
var import_Button = require("../../buttons/button/Button.js");
var import_useId = require("../../core/hooks/useId.js");
var import_useSlots = require("../../core/hooks/useSlots.js");
var import_Flex = require("../../layouts/flex/Flex.js");
var import_field_css = require("../../styles/field.sty.js");
const AccordionSection = /* @__PURE__ */ (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const {
      disabled = false,
      id: idProp,
      sectionId,
      children,
      className: consumerClassName,
      style: consumerStyle,
      keepMounted = false,
      "data-testid": dataTestId,
      as,
      ...remainingProps
    } = props;
    const [slots] = (0, import_useSlots.useSlots)(children, {
      label: import_AccordionSectionLabel.AccordionSectionLabel,
      content: import_AccordionSectionContent.AccordionSectionContent
    });
    const { triggerPosition, interactive, color, size, state, dispatch } = (0, import_AccordionContext.useAccordionContext)();
    const sectionColor = props.color ?? color;
    const triggerId = (0, import_useId.useId)();
    const contentId = (0, import_useId.useId)();
    const triggerIsStart = triggerPosition === "start";
    const id = sectionId ?? idProp;
    const isExpanded = state.expanded.has(id);
    const labelId = (0, import_useId.useId)();
    const containerRef = (0, import_react.useRef)(null);
    const { width } = (0, import_use_resize_observer.default)({ ref: containerRef.current });
    const [offsetWidth, setOffsetWidth] = (0, import_react.useState)(0);
    const updateOverflowWidth = (0, import_react.useCallback)(() => {
      if (containerRef.current) {
        setOffsetWidth(
          containerRef.current.scrollWidth - containerRef.current.offsetWidth
        );
      }
    }, []);
    const dataDtProperties = `state:${isExpanded ? "collapse" : "expand"}`;
    (0, import_react.useLayoutEffect)(() => {
      updateOverflowWidth();
    }, [updateOverflowWidth, width]);
    const trigger = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Button.Button,
      {
        id: triggerId,
        onClick: () => {
          if (!interactive) {
            dispatch({ type: import_accordion_state.AccordionActions.TOGGLE_SECTION, payload: id });
          }
        },
        color: sectionColor,
        size,
        disabled,
        "aria-labelledby": labelId,
        "aria-expanded": isExpanded,
        "aria-controls": isExpanded || keepMounted ? contentId : void 0,
        "data-dt-name": "ExpandCollapse",
        "data-dt-properties": dataDtProperties,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_strato_icons.ChevronDownSmallIcon,
          {
            className: (0, import_clsx.default)(import_Accordion_css.accordionTriggerIconCSS, {
              [import_Accordion_css.accordionTriggerIconStartCSS]: triggerIsStart,
              [import_Accordion_css.accordionTriggerIconEndCSS]: !triggerIsStart,
              [import_Accordion_css.accordionTriggerIconExpandedCSS]: isExpanded
            })
          }
        ) })
      }
    );
    const Section = as ?? "div";
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      Section,
      {
        "data-testid": dataTestId,
        ref: forwardedRef,
        className: (0, import_clsx.default)(
          (0, import_field_css.fieldTokens)({ color: sectionColor }),
          consumerClassName
        ),
        style: consumerStyle,
        ...remainingProps,
        children: (slots.label || slots.content) && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
          import_AccordionSectionContext.AccordionSectionContext.Provider,
          {
            value: {
              labelId,
              contentId,
              isExpanded,
              keepMounted,
              size
            },
            children: [
              /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
                import_Flex.Flex,
                {
                  ref: containerRef,
                  flexDirection: "row",
                  alignItems: "center",
                  justifyContent: triggerIsStart ? void 0 : "space-between",
                  gap: size === "default" ? 8 : 4,
                  id: labelId,
                  onClick: () => {
                    if (interactive && !disabled) {
                      dispatch({
                        type: import_accordion_state.AccordionActions.TOGGLE_SECTION,
                        payload: id
                      });
                    }
                  },
                  style: {
                    cursor: interactive && !disabled ? "pointer" : void 0,
                    width: `calc(100% - ${offsetWidth}px)`
                  },
                  "data-dt-name": "ExpandCollapseHeader",
                  "data-dt-properties": dataDtProperties,
                  children: [
                    triggerIsStart && trigger,
                    slots.label,
                    !triggerIsStart && trigger
                  ]
                }
              ),
              slots.content
            ]
          }
        )
      }
    );
  }
);
AccordionSection.displayName = "Accordion.Section";
