// @dynatrace/strato-components/content/accordion/AccordionSectionContent.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var AccordionSectionContent_exports = {};
__export(AccordionSectionContent_exports, {
  AccordionSectionContent: () => AccordionSectionContent
});
module.exports = __toCommonJS(AccordionSectionContent_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_framer_motion = require("framer-motion");
var import_react = require("react");
var import_animations_js = __toESM(require("@dynatrace/strato-design-tokens/animations-js"));
var import_Accordion_css = require("./Accordion.sty.js");
var import_AccordionSectionContext = require("./contexts/AccordionSectionContext.js");
var import_can_receive_default_text_format = require("../../core/utils/can-receive-default-text-format.js");
var import_framer_motion_easings = require("../../core/utils/framer-motion-easings.js");
var import_Text = require("../../typography/text/Text.js");
const animationVariantSectionContent = {
  expanded: {
    height: "auto",
    opacity: 1
  },
  collapsed: {
    height: 0,
    opacity: 0
  }
};
const AccordionSectionContent = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    children,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId
  } = props;
  const { labelId, contentId, isExpanded, keepMounted, size } = (0, import_react.useContext)(
    import_AccordionSectionContext.AccordionSectionContext
  );
  const childIsText = (0, import_can_receive_default_text_format.canReceiveDefaultTextFormat)(children);
  const [isAnimating, setIsAnimating] = (0, import_react.useState)(false);
  const shouldReduceMotion = (0, import_framer_motion.useReducedMotion)();
  const content = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_framer_motion.motion.div,
    {
      "data-testid": dataTestId,
      variants: animationVariantSectionContent,
      initial: "collapsed",
      animate: isExpanded ? "expanded" : "collapsed",
      exit: "collapsed",
      transition: {
        duration: shouldReduceMotion ? 0 : import_animations_js.default.ExpandCollapse.Duration.Base,
        ease: import_framer_motion_easings.expandCollapseEasing
      },
      className: (0, import_clsx.default)(import_Accordion_css.accordionContentWrapperCSS, consumerClassName),
      style: {
        ...consumerStyle,
        willChange: "initial",
        overflow: keepMounted && (isAnimating || !isExpanded) ? "hidden" : void 0
      },
      onAnimationStart: () => {
        setIsAnimating(true);
      },
      onAnimationComplete: () => {
        setIsAnimating(false);
      },
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        "div",
        {
          className: (0, import_clsx.default)([
            (0, import_Accordion_css.accordionContentCSS)({
              size
            }),
            consumerClassName
          ]),
          ref: forwardedRef,
          id: contentId,
          role: "region",
          "aria-labelledby": labelId,
          children: childIsText ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { textStyle: "base", as: "p", children }) : children
        }
      )
    }
  );
  if (keepMounted) {
    return content;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { style: { overflow: isAnimating ? "hidden" : void 0 }, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_framer_motion.AnimatePresence, { initial: false, children: isExpanded && content }) });
});
AccordionSectionContent.displayName = "Accordion.SectionContent";
