// @dynatrace/strato-components/content/accordion/Accordion.sty.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Accordion_css_exports = {};
__export(Accordion_css_exports, {
  accordionContentCSS: () => accordionContentCSS,
  accordionContentWrapperCSS: () => accordionContentWrapperCSS,
  accordionDividerMarginCSS: () => accordionDividerMarginCSS,
  accordionLabelCSS: () => accordionLabelCSS,
  accordionTriggerIconCSS: () => accordionTriggerIconCSS,
  accordionTriggerIconEndCSS: () => accordionTriggerIconEndCSS,
  accordionTriggerIconExpandedCSS: () => accordionTriggerIconExpandedCSS,
  accordionTriggerIconStartCSS: () => accordionTriggerIconStartCSS
});
module.exports = __toCommonJS(Accordion_css_exports);
var import_AppRoot_css_ts_vanilla = require("../../core/components/app-root/AppRoot.css");
var import_field_css_ts_vanilla = require("../../styles/field.css");
var import_Accordion_css_ts_vanilla = require("./Accordion.css");
var import_createRuntimeFn = require("@vanilla-extract/recipes/createRuntimeFn");
var accordionContentCSS = (0, import_createRuntimeFn.createRuntimeFn)({ defaultClassName: "_12zc70k8-3-6-0", variantClassNames: { size: { "default": "_12zc70k9-3-6-0", condensed: "_12zc70ka-3-6-0" } }, defaultVariants: {}, compoundVariants: [] });
var accordionContentWrapperCSS = "_12zc70k7-3-6-0";
var accordionDividerMarginCSS = { "default": "_12zc70k1-3-6-0", condensed: "_12zc70k2-3-6-0" };
var accordionLabelCSS = "_12zc70k0-3-6-0";
var accordionTriggerIconCSS = "_12zc70k3-3-6-0";
var accordionTriggerIconEndCSS = "_12zc70k5-3-6-0";
var accordionTriggerIconExpandedCSS = "_12zc70k6-3-6-0";
var accordionTriggerIconStartCSS = "_12zc70k4-3-6-0";
