// @dynatrace/strato-components/forms/calendar/utils/helpers.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var helpers_exports = {};
__export(helpers_exports, {
  convertToCalendarStructure: () => convertToCalendarStructure,
  getDays: () => getDays,
  getHighlightingProps: () => getHighlightingProps,
  getTimeOffsetFromInitialValue: () => getTimeOffsetFromInitialValue,
  getWeekStartOrEnd: () => getWeekStartOrEnd
});
module.exports = __toCommonJS(helpers_exports);
var import_tz = require("@date-fns/tz");
var import_date_fns = require("date-fns");
var import_weekstart = require("weekstart");
var import_user_preferences = require("@dynatrace-sdk/user-preferences");
var import_helpers = require("../../../core/time/helpers.js");
var import_helpers2 = require("../../forms-core/time/helpers.js");
var import_timeParser = require("../../forms-core/time/timeParser.js");
var import_calendarConfiguration = require("../types/calendarConfiguration.js");
function getDays(activeDate, weekStartsOn) {
  const monthStart = (0, import_date_fns.startOfMonth)(activeDate);
  const firstWeekStart = (0, import_date_fns.startOfWeek)(monthStart, {
    weekStartsOn
  });
  const monthEnd = (0, import_date_fns.lastDayOfMonth)(monthStart);
  const lastWeekEnd = (0, import_date_fns.lastDayOfWeek)(monthEnd, { weekStartsOn });
  let prevMonthDays = getSiblingMonthDays(monthStart, firstWeekStart, -1);
  const currentMonthDays = toDayArray(
    (0, import_helpers2.eachDayOfInterval)({
      start: monthStart,
      end: monthEnd
    }),
    true
  );
  const nextMonthDays = getSiblingMonthDays(monthEnd, lastWeekEnd, 1);
  const days = [...prevMonthDays, ...currentMonthDays, ...nextMonthDays];
  while (days.length < import_calendarConfiguration.NUMBER_OF_WEEKDAYS * import_calendarConfiguration.NUMBER_OF_WEEK_ROWS) {
    if (!prevMonthDays.length) {
      prevMonthDays = toDayArray(
        (0, import_helpers2.eachDayOfInterval)({
          // Typecast to TZDate since TypeScript doesn't infer addWeeks return type directly
          start: (0, import_date_fns.addWeeks)(firstWeekStart, -1),
          end: (0, import_date_fns.addDays)(firstWeekStart, -1)
        })
      );
      days.unshift(...prevMonthDays);
    } else {
      days.push(
        ...toDayArray(
          (0, import_helpers2.eachDayOfInterval)({
            // Typecast to TZDate since TypeScript doesn't infer addWeeks return type directly
            start: (0, import_date_fns.addDays)(lastWeekEnd, 1),
            end: (0, import_date_fns.addWeeks)(lastWeekEnd, 1)
          })
        )
      );
    }
  }
  return days;
}
function getSiblingMonthDays(monthStartEnd, weekStartEnd, direction) {
  if ((0, import_date_fns.isEqual)(monthStartEnd, weekStartEnd)) {
    return [];
  }
  const siblingMonthStartEnd = (0, import_date_fns.addDays)(monthStartEnd, direction);
  const sortedDays = [siblingMonthStartEnd, weekStartEnd].sort(
    (a, b) => (0, import_date_fns.compareAsc)(a, b)
  );
  return toDayArray(
    (0, import_helpers2.eachDayOfInterval)({
      start: sortedDays[0],
      end: sortedDays[1]
    })
  );
}
function toDayArray(array, isCurrentMonthDay = false) {
  return array.map(
    (date) => ({
      date,
      isCurrentMonthDay
    })
  );
}
function convertToCalendarStructure(type = "single", value, min, max, precision) {
  if (type === "single" && !value) {
    return null;
  }
  if (typeof value === "string" || value instanceof Date) {
    const stringDate = typeof value === "string" ? value : value.toISOString();
    const timeOffset = getTimeOffsetFromInitialValue(
      stringDate,
      "single",
      min,
      max,
      precision
    );
    return (0, import_helpers2.getTimeValueFromSimpleValue)(
      value,
      timeOffset,
      void 0,
      void 0,
      min,
      max,
      precision
    ) ?? null;
  }
  if ((0, import_helpers.isTimeValue)(value)) {
    return (0, import_timeParser.removePrecisionFromTimeValue)(value, precision);
  }
  if (value) {
    const timeframe = (0, import_timeParser.mapToPartialTimeframe)(value);
    const timeOffset = getTimeOffsetFromInitialValue(
      timeframe,
      "range",
      min,
      max,
      precision
    );
    const from = (0, import_helpers2.getTimeValueFromSimpleValue)(
      timeframe?.from?.value,
      timeOffset,
      "from",
      timeframe?.from?.type === "expression",
      min,
      void 0,
      precision
    );
    const to = (0, import_helpers2.getTimeValueFromSimpleValue)(
      timeframe?.to?.value,
      timeOffset,
      "to",
      timeframe?.to?.type === "expression",
      void 0,
      max,
      precision
    );
    return (0, import_helpers2.sortTimeframeValues)({ from, to }, min, max, precision);
  }
  return { from: null, to: null };
}
function getRangePosition(newDate, from, to) {
  if (from && (0, import_date_fns.isSameDay)(newDate, from)) {
    return "first";
  }
  if (to && (0, import_date_fns.isSameDay)(newDate, to)) {
    return "last";
  }
  if ((0, import_date_fns.isWithinInterval)(newDate, {
    start: (0, import_date_fns.startOfDay)(from),
    end: (0, import_date_fns.endOfDay)(to)
  })) {
    return "inbetween";
  }
}
function isRelativeDate(highlightType, selectedDate) {
  switch (highlightType) {
    case "first": {
      if (selectedDate.from) {
        return selectedDate.from.type === "iso8601" ? "absolute" : "relative";
      }
      return void 0;
    }
    case "last": {
      if (selectedDate.to) {
        return selectedDate.to.type === "iso8601" ? "absolute" : "relative";
      }
      if (!selectedDate.to && selectedDate.from) {
        return selectedDate.from.type === "iso8601" ? "absolute" : "relative";
      }
      return void 0;
    }
    case "same": {
      if (selectedDate.from?.type === "expression" && selectedDate.to?.type === "expression") {
        return "relative";
      }
      if (selectedDate.from?.type === "expression" && selectedDate.to?.type === "iso8601") {
        return "relative-absolute";
      }
      if (selectedDate.from?.type === "iso8601" && selectedDate.to?.type === "expression") {
        return "absolute-relative";
      }
      return "absolute";
    }
    case "inbetween":
    default: {
      return void 0;
    }
  }
}
function getHighlightType(from, to, newDate) {
  if (to && from && (0, import_date_fns.isSameDay)(from, newDate) && (0, import_date_fns.isSameDay)(to, newDate) || !to && from && (0, import_date_fns.isSameDay)(from, newDate) || !from && to && (0, import_date_fns.isSameDay)(to, newDate)) {
    return "same";
  }
  if (from && to && (0, import_date_fns.isBefore)(from, to)) {
    return getRangePosition(newDate, from, to);
  }
  if (from && to && (0, import_date_fns.isBefore)(to, from)) {
    return getRangePosition(newDate, to, from);
  }
  return void 0;
}
const getHighlightPropsFromTimeframe = (date, selectedDate, hoveredDate) => {
  const { from: selectedFrom, to: selectedTo } = selectedDate;
  if (!selectedFrom && !selectedTo) {
    return {};
  }
  const relativeDate = import_tz.TZDateMini.tz((0, import_user_preferences.getTimezone)()).getTime();
  const from = (0, import_helpers2.timeValueToDate)(selectedFrom, relativeDate);
  const to = (0, import_helpers2.timeValueToDate)(selectedTo, relativeDate);
  const isSelected = !!(from && (0, import_date_fns.isSameDay)(date, from) || to && (0, import_date_fns.isSameDay)(date, to));
  const safeFromDate = from ?? hoveredDate;
  const safeToDate = to ?? hoveredDate;
  const highlightType = getHighlightType(safeFromDate, safeToDate, date);
  const isEntireRangeSelected = !!selectedFrom && !!selectedTo && !!highlightType;
  const highlightColor = isSelected || isEntireRangeSelected ? "primary" : "neutral";
  const isRelative = isSelected ? isRelativeDate(highlightType, selectedDate) : void 0;
  return {
    highlightType,
    highlightColor,
    isSelected,
    isIncompleteSelection: !isEntireRangeSelected,
    selectionVariant: isRelative
  };
};
const getHighlightingProps = (date, selectedDate, type, hoveredDate = null) => {
  if (!selectedDate) {
    return {};
  }
  if ((0, import_helpers.isTimeframe)(selectedDate, type)) {
    return getHighlightPropsFromTimeframe(date, selectedDate, hoveredDate);
  }
  if ((0, import_helpers.isTimeValue)(selectedDate)) {
    const newSelectedDate = (0, import_helpers2.timeValueToDate)(selectedDate, Date.now());
    if (!newSelectedDate) {
      return {};
    }
    const isSelected = (0, import_date_fns.isSameDay)(date, newSelectedDate);
    if (isSelected) {
      return {
        isSelected: true,
        highlightType: "same",
        highlightColor: "primary",
        isIncompleteSelection: false,
        selectionVariant: selectedDate.type === "expression" ? "relative" : "absolute"
      };
    }
    return {
      isSelected: false
    };
  }
  return {};
};
function getWeekStartOrEnd(indicator, activeDate) {
  const region = (0, import_user_preferences.getRegionalFormat)();
  const weekStartsOn = (0, import_weekstart.getWeekStartByLocale)(region);
  const start = indicator === "startOfWeek" ? (0, import_date_fns.startOfWeek)(activeDate, { weekStartsOn }) : (0, import_date_fns.endOfWeek)(activeDate, { weekStartsOn });
  return ((0, import_date_fns.intervalToDuration)({
    start,
    end: activeDate
  }).days ?? 0) * -1;
}
function getTimeOffsetFromInitialValue(value, type = "single", min = null, max = null, precision = "milliseconds") {
  if ((0, import_helpers.isTimeframe)(value, type)) {
    if (!value) {
      return { from: "startOfDay", to: "endOfDay" };
    }
    let from = value["from"]?.absoluteDate;
    let to = value["to"]?.absoluteDate;
    const configuredTimezone = (0, import_user_preferences.getTimezone)();
    const fromIsAtStartOfDay = from && from === (0, import_date_fns.startOfDay)(new import_tz.TZDateMini(from, configuredTimezone)).toISOString();
    const fromIsMin = min && from && (0, import_timeParser.removePrecision)(from, precision) === (0, import_timeParser.removePrecision)(min?.absoluteDate, precision);
    if (!from || fromIsAtStartOfDay || value["from"]?.type === "expression" || fromIsMin) {
      from = "startOfDay";
    }
    const toIsAtEndOfDay = to && to === (0, import_date_fns.endOfDay)(new import_tz.TZDateMini(to, configuredTimezone)).toISOString();
    const toIsMax = max && to && (0, import_timeParser.removePrecision)(to, precision) === (0, import_timeParser.removePrecision)(max?.absoluteDate, precision);
    if (!to || toIsAtEndOfDay || value.to?.type === "expression" || toIsMax) {
      to = "endOfDay";
    }
    return {
      from,
      to
    };
  }
  if (precision === "day") {
    return "startOfDay";
  }
  const offset = (0, import_helpers.isTimeValue)(value) ? value.value : value;
  if (typeof offset === "string") {
    return (0, import_timeParser.isExpressionString)(offset) ? "startOfDay" : offset;
  }
  return "startOfDay";
}
