// @dynatrace/strato-components/esm/overlays/hooks/useOverlayWithTrigger.js
import { eventListeners } from "@popperjs/core";
import {
  useCallback
} from "react";
import { usePopper } from "react-popper";
import { useDialog } from "./useDialog.js";
import { useKeyboardInteractionOutside } from "./useKeyboardInteractionOutside.js";
import {
  useOverlay,
  useOverlayTriggerState
} from "./useOverlay.js";
import { usePress } from "./usePress.js";
import { useStateRef } from "./useStateRef.js";
import { useId } from "../../core/hooks/useId.js";
import {
  useOverlayStacking
} from "../../core/hooks/useOverlayStacking.js";
import { getFirstFocusableChild as _getFirstFocusableChild } from "../../core/utils/focus-management/get-first-focusable-child.js";
import { getLastFocusableChild as _getLastFocusableChild } from "../../core/utils/focus-management/get-last-focusable-child.js";
import { mergeProps as _mergeProps } from "../../core/utils/merge-props.js";
import { isFormControlRef } from "../../forms/shared-types.js";
import {
  observeAnimationDone,
  observeOverlaySize,
  overlayHeightStrategyHandling,
  overlayWidthStrategyHandling
} from "../utils/popper-modifiers.js";
function getAriaHasPopup(type) {
  if (type === "menu") {
    return true;
  } else if (type === "listbox") {
    return "listbox";
  }
  return void 0;
}
function useOverlayWithTrigger(props = {}) {
  const {
    placement = "bottom",
    placementFallbacks = ["top"],
    offset = 0,
    crossOffset = 0,
    additionalElevation,
    disabled = false,
    widthStrategy = "content",
    heightStrategy = "content",
    trapFocus = false,
    defaultOpen = false,
    open,
    onOpenChange,
    shouldCloseOnBlur = false,
    "aria-labelledby": ariaLabelledBy
  } = props;
  const [targetElement, setTargetElement] = useStateRef();
  const [overlayElement, setOverlayElement, overlayRef] = useStateRef();
  const [, setTriggerElement, triggerRef] = useStateRef();
  const overlayContainerProps = useOverlayStacking(
    void 0,
    additionalElevation
  );
  const overlayState = useOverlayTriggerState({
    defaultOpen,
    isOpen: open,
    onOpenChange
  });
  const { isOpen, close, toggle } = overlayState;
  const overlayId = useId();
  const onKeyDown = useCallback(
    (e) => {
      if (e.key === "Tab" && isOpen) {
        const overlay = document.getElementById(`${overlayId}`);
        if (overlay) {
          const firstFocusableChild = _getFirstFocusableChild(overlay);
          const lastFocusableChild = _getLastFocusableChild(overlay);
          if (!firstFocusableChild) {
            e.preventDefault();
            close();
            return;
          }
          const isFirstFocusable = firstFocusableChild === e.target;
          const isLastFocusable = lastFocusableChild === e.target;
          if (!trapFocus && (e.shiftKey && isFirstFocusable || !e.shiftKey && isLastFocusable)) {
            e.preventDefault();
            close();
          }
        }
      }
    },
    [close, isOpen, overlayId, trapFocus]
  );
  const triggerProps = {
    "aria-haspopup": getAriaHasPopup("dialog"),
    "aria-expanded": isOpen,
    "aria-controls": isOpen ? overlayId : void 0,
    onKeyDown
  };
  const { pressProps } = usePress({
    onPress: toggle,
    isDisabled: disabled
  });
  const { styles: positioningStyles, attributes: positioningAttributes } = usePopper(targetElement, overlayElement, {
    placement,
    modifiers: [
      {
        name: "offset",
        options: {
          offset: [crossOffset, offset]
        }
      },
      {
        name: "flip",
        options: {
          fallbackPlacements: [placement, ...placementFallbacks]
        }
      },
      overlayWidthStrategyHandling,
      {
        name: "overlayWidthStrategyHandling",
        enabled: widthStrategy !== "content",
        options: { widthStrategy }
      },
      overlayHeightStrategyHandling,
      {
        name: "overlayHeightStrategyHandling",
        enabled: heightStrategy !== "content",
        options: { heightStrategy }
      },
      observeOverlaySize,
      observeAnimationDone,
      eventListeners
    ]
  });
  const { overlayProps } = useOverlay(
    {
      isOpen,
      isDismissible: true,
      shouldCloseOnInteractOutside: (element) => {
        const triggerElement = isFormControlRef(triggerRef) ? triggerRef.element : triggerRef;
        if (triggerElement === null) {
          return true;
        }
        return element !== triggerRef.current && triggerRef.current?.contains(element) === false;
      },
      onClose: () => {
        close();
      },
      shouldCloseOnBlur
    },
    overlayRef
  );
  const { dialogProps } = useDialog({});
  const { "aria-labelledby": _unused, ...dialogPropsWithoutLabelledBy } = dialogProps;
  useKeyboardInteractionOutside({
    ref: overlayRef,
    isDisabled: !isOpen,
    onInteractOutside: (event) => {
      if (event.key === "Escape") {
        close();
      }
    }
  });
  const updateTriggerElement = useCallback(
    (triggerRef2) => {
      const newTriggerElement = isFormControlRef(triggerRef2) ? triggerRef2.element : triggerRef2;
      setTriggerElement(newTriggerElement);
    },
    [setTriggerElement]
  );
  const updateTargetElement = useCallback(
    (targetRef) => {
      const newTargetElement = isFormControlRef(targetRef) ? targetRef.element : targetRef;
      setTargetElement(newTargetElement);
    },
    [setTargetElement]
  );
  return {
    triggerRef: updateTriggerElement,
    targetRef: updateTargetElement,
    overlayRef: setOverlayElement,
    overlayProps: {
      ..._mergeProps(
        overlayProps,
        { ...dialogPropsWithoutLabelledBy, "aria-labelledby": ariaLabelledBy },
        // we also need to add the onKeyDown handler in here for closing overlays with interactive content when tabbing and shouldCloseOnBlur is false (default)
        // if shouldCloseOnBlur is set to true, the focus handlers from the returned useOverlay overlayProps take care of the closing
        // if there's no interactive content, the trigger keeps the focus, so the onKeyDown event from the triggerProps takes care of closing
        { onKeyDown: shouldCloseOnBlur ? void 0 : onKeyDown },
        { style: positioningStyles.popper, id: overlayId, trapFocus },
        positioningAttributes.popper
      )
    },
    overlayContainerProps,
    overlayState,
    overlayTriggerProps: _mergeProps(pressProps, triggerProps, {
      type: "button",
      disabled
    })
  };
}
export {
  useOverlayWithTrigger
};
