// @dynatrace/strato-components/esm/overlays/hooks/useOverlay.js
import { jsx } from "react/jsx-runtime";
import {
  useCallback,
  useContext,
  useEffect,
  useRef
} from "react";
import { VisuallyHidden } from "../../core/components/visually-hidden/VisuallyHidden.js";
import { OverlayStackingContext } from "../../core/contexts/OverlayStacking.js";
import { useControllableState } from "../../core/hooks/useControllableState.js";
const visibleOverlays = [];
const RADIX_MENU_CONTENT_SELECTOR = "[data-radix-menu-content]";
const overlayStackListeners = /* @__PURE__ */ new Set();
function notifyOverlayStackListeners() {
  overlayStackListeners.forEach((listener) => listener());
}
function subscribeToOverlayStack(listener) {
  overlayStackListeners.add(listener);
  return () => {
    overlayStackListeners.delete(listener);
  };
}
function getOverlayStackSize() {
  return visibleOverlays.length;
}
function closeAllOverlays(level = 0) {
  if (visibleOverlays.length === 0) {
    return;
  }
  for (let i = visibleOverlays.length - 1; i >= 0; i--) {
    const overlay = visibleOverlays[i];
    if (overlay.stackingLevel >= level) {
      overlay.close();
    } else {
      break;
    }
  }
}
function useOverlay(props, ref) {
  const {
    onClose,
    shouldCloseOnBlur,
    isOpen,
    isDismissible = false,
    isKeyboardDismissDisabled = false,
    shouldCloseOnInteractOutside
  } = props;
  const overlayStacking = useContext(OverlayStackingContext);
  const hasPendingCloseRef = useRef(false);
  const onCloseRef = useRef(onClose);
  useEffect(() => {
    onCloseRef.current = onClose;
  }, [onClose]);
  useEffect(() => {
    if (!isOpen) {
      hasPendingCloseRef.current = false;
    }
  }, [isOpen]);
  const requestClose = useCallback(
    (reason, onlyIfTopMost = true) => {
      if (!onCloseRef.current || !isOpen || hasPendingCloseRef.current) {
        return;
      }
      if (onlyIfTopMost) {
        const topEntry = visibleOverlays[visibleOverlays.length - 1];
        if (topEntry?.ref !== ref) {
          return;
        }
      }
      hasPendingCloseRef.current = true;
      onCloseRef.current(reason);
    },
    [isOpen, ref]
  );
  useEffect(() => {
    if (isOpen && onCloseRef.current) {
      const entry = {
        ref,
        stackingLevel: overlayStacking,
        close: () => requestClose("programmatic", false)
      };
      visibleOverlays.push(entry);
      notifyOverlayStackListeners();
      return () => {
        const index = visibleOverlays.findIndex((e) => e.ref === ref);
        if (index >= 0) {
          visibleOverlays.splice(index, 1);
          notifyOverlayStackListeners();
        }
      };
    }
    return () => {
      const index = visibleOverlays.findIndex((e) => e.ref === ref);
      if (index >= 0) {
        visibleOverlays.splice(index, 1);
      }
    };
  }, [isOpen, ref, overlayStacking, requestClose]);
  const onInteractOutside = (e) => {
    if (!shouldCloseOnInteractOutside || shouldCloseOnInteractOutside(e.target)) {
      requestClose("backdrop");
    }
  };
  const onInteractOutsideStart = (e) => {
    if (!isDismissible || !shouldCloseOnBlur) {
      return;
    }
    if (!shouldCloseOnInteractOutside || shouldCloseOnInteractOutside(e.target)) {
      requestClose("blur");
    }
  };
  const onKeyDown = useCallback(
    (event) => {
      if (event.key === "Escape" && !isKeyboardDismissDisabled) {
        event.preventDefault();
        requestClose("escape");
      }
      event.stopPropagation();
    },
    [isKeyboardDismissDisabled, requestClose]
  );
  useInteractOutside({
    ref,
    onInteractOutside: isDismissible ? onInteractOutside : void 0,
    onInteractOutsideStart
  });
  const { focusWithinProps } = useFocusWithin({
    isDisabled: !shouldCloseOnBlur,
    onBlurWithin: (e) => {
      if (!shouldCloseOnInteractOutside || shouldCloseOnInteractOutside(e.relatedTarget)) {
        requestClose("blur");
      }
    }
  });
  return {
    overlayProps: {
      onKeyDown,
      ...focusWithinProps
    }
  };
}
function useOverlayTriggerState(props) {
  const onOpenChange = props.onOpenChange ?? ((isOpen2) => {
  });
  const [isOpenValue, setOpen] = useControllableState({
    defaultValue: props.defaultOpen || false,
    value: props.isOpen,
    onChange: onOpenChange
  });
  const isOpen = isOpenValue ?? false;
  return {
    isOpen,
    open: useCallback(() => setOpen(true), [setOpen]),
    close: useCallback(() => setOpen(false), [setOpen]),
    toggle: useCallback(() => setOpen(!isOpen), [isOpen, setOpen])
  };
}
function useInteractOutside(props) {
  const { ref, onInteractOutside, isDisabled, onInteractOutsideStart } = props;
  const stateRef = useRef({
    isPointerDown: false,
    ignoreEmulatedMouseEvents: false,
    onInteractOutside,
    onInteractOutsideStart
  });
  const state = stateRef.current;
  state.onInteractOutside = onInteractOutside;
  state.onInteractOutsideStart = onInteractOutsideStart;
  useEffect(() => {
    if (isDisabled) {
      return;
    }
    const onPointerDown = (e) => {
      if (isValidEvent(e, ref) && state.onInteractOutside) {
        if (state.onInteractOutsideStart) {
          state.onInteractOutsideStart(e);
        }
        state.isPointerDown = true;
      }
    };
    if (typeof PointerEvent !== "undefined") {
      const onPointerUp = (e) => {
        if (state.isPointerDown && state.onInteractOutside && isValidEvent(e, ref)) {
          state.isPointerDown = false;
          state.onInteractOutside(e);
        }
      };
      document.body.addEventListener("pointerdown", onPointerDown);
      document.body.addEventListener("pointerup", onPointerUp);
      return () => {
        document.body.removeEventListener("pointerdown", onPointerDown);
        document.body.removeEventListener("pointerup", onPointerUp);
      };
    } else {
      const onMouseUp = (e) => {
        if (state.ignoreEmulatedMouseEvents) {
          state.ignoreEmulatedMouseEvents = false;
        } else if (state.isPointerDown && state.onInteractOutside && isValidEvent(e, ref)) {
          state.isPointerDown = false;
          state.onInteractOutside(e);
        }
      };
      const onTouchEnd = (e) => {
        state.ignoreEmulatedMouseEvents = true;
        if (state.onInteractOutside && state.isPointerDown && isValidEvent(e, ref)) {
          state.isPointerDown = false;
          state.onInteractOutside(e);
        }
      };
      document.body.addEventListener("mousedown", onPointerDown);
      document.body.addEventListener("mouseup", onMouseUp);
      document.body.addEventListener("touchstart", onPointerDown);
      document.body.addEventListener("touchend", onTouchEnd);
      return () => {
        document.body.removeEventListener("mousedown", onPointerDown);
        document.body.removeEventListener("mouseup", onMouseUp);
        document.body.removeEventListener("touchstart", onPointerDown);
        document.body.removeEventListener("touchend", onTouchEnd);
      };
    }
  }, [ref, state, isDisabled]);
}
function isValidEvent(event, ref) {
  if (event.button > 0) {
    return false;
  }
  if (event.target) {
    const ownerDocument = event.target.ownerDocument;
    if (!ownerDocument || !ownerDocument.documentElement.contains(event.target)) {
      return false;
    }
  }
  if (event.target instanceof Element && event.target.closest(RADIX_MENU_CONTENT_SELECTOR)) {
    return false;
  }
  return !!ref.current && !ref.current.contains(event.target);
}
function useFocusWithin(props) {
  const state = useRef({
    isFocusWithin: false
  }).current;
  if (props.isDisabled) {
    return { focusWithinProps: {} };
  }
  const onFocus = (e) => {
    if (!state.isFocusWithin) {
      if (props.onFocusWithin) {
        props.onFocusWithin(e);
      }
      if (props.onFocusWithinChange) {
        props.onFocusWithinChange(true);
      }
      state.isFocusWithin = true;
    }
  };
  const onBlur = (e) => {
    const currentTarget = e.currentTarget;
    const relatedTarget = e.relatedTarget;
    if (state.isFocusWithin && !currentTarget.contains(relatedTarget)) {
      if (props.onBlurWithin) {
        props.onBlurWithin(e);
      }
      if (props.onFocusWithinChange) {
        props.onFocusWithinChange(false);
      }
      state.isFocusWithin = false;
    }
  };
  return {
    focusWithinProps: {
      onFocus,
      onBlur
    }
  };
}
function DismissButton(props) {
  const { onDismiss } = props;
  const onClick = () => {
    if (onDismiss) {
      onDismiss();
    }
  };
  return /* @__PURE__ */ jsx(VisuallyHidden, { children: /* @__PURE__ */ jsx("button", { tabIndex: -1, "aria-label": "dismiss", onClick }) });
}
export {
  DismissButton,
  closeAllOverlays,
  getOverlayStackSize,
  subscribeToOverlayStack,
  useOverlay,
  useOverlayTriggerState
};
