// @dynatrace/strato-components/charts/timeseries/components/displays/line/LineChartDisplay.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var LineChartDisplay_exports = {};
__export(LineChartDisplay_exports, {
  LineChartDisplay: () => LineChartDisplay
});
module.exports = __toCommonJS(LineChartDisplay_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_accessors = require("./accessors.js");
var import_LineCanvas = require("../../../../core/components/canvas/components/LineCanvas.js");
var import_display_points = require("../../../../core/components/canvas/utils/display-points.js");
var import_useGridPlotSize = require("../../../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_is_isolated_point = require("../../../../core/utils/is-isolated-point.js");
var import_safe_get = require("../../../../core/utils/safe-get.js");
var import_constants = require("../../../constants.js");
var import_useSeriesCurve = require("../../../hooks/useSeriesCurve.js");
var import_useTimeseriesFilteredData = require("../../../hooks/useTimeseriesFilteredData.js");
var import_useYScales = require("../../../hooks/useYScales.js");
var import_clamp_value = require("../shared/utils/clamp-value.js");
const LineChartDisplay = ({ series, xScale }) => {
  const yScales = (0, import_useYScales.useYScales)();
  const { seriesMap, datapointsMap } = (0, import_useTimeseriesFilteredData.useTimeseriesFilteredData)();
  const curve = (0, import_useSeriesCurve.useSeriesCurve)();
  const { width, height } = (0, import_useGridPlotSize.useGridPlotSize)();
  const lineSeries = (0, import_react.useMemo)(() => [...series].reverse(), [series]);
  const isolatedAccessor = (displayPoints) => (datapoints, index) => displayPoints || (0, import_is_isolated_point.isIsolatedPoint)(
    datapoints,
    index,
    (point) => (0, import_accessors.definedAccessor)(point, datapointsMap)
  );
  const clampedScale = (0, import_clamp_value.getClampedTimeScale)(xScale);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("figure", { "aria-label": import_constants.LINE_CHART_DISPLAY_ARIA_LABEL, children: lineSeries.map((series2) => {
    const seriesMetadata = (0, import_safe_get.safeGet)(seriesMap, series2);
    const displayPoints = (0, import_display_points.shouldDisplayPoints)(
      seriesMetadata.pointsDisplay,
      width,
      series2.datapoints.length
    );
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_LineCanvas.LineCanvas,
      {
        data: series2.datapoints,
        canvasWidth: width,
        canvasHeight: height,
        x: (0, import_accessors.xAxisAccessor)(xScale, datapointsMap, clampedScale),
        y: (0, import_accessors.yAxisAccessor)(yScales, datapointsMap, seriesMetadata.unit),
        defined: (point) => (0, import_accessors.definedAccessor)(point, datapointsMap),
        displayScatterPlot: isolatedAccessor(displayPoints),
        color: () => seriesMetadata.color.Default,
        opacity: 1,
        curve,
        enableDotOverflow: true
      },
      seriesMetadata.id
    );
  }) });
};
LineChartDisplay["displayName"] = "LineChartDisplay";
