// @dynatrace/strato-components/charts/gantt-chart/constants/defaults.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var defaults_exports = {};
__export(defaults_exports, {
  DEFAULT_BAR_OPACITY: () => DEFAULT_BAR_OPACITY,
  DEFAULT_GANTT_COLORS: () => DEFAULT_GANTT_COLORS,
  DEFAULT_GANTT_LINE_SIZE: () => DEFAULT_GANTT_LINE_SIZE,
  DEFAULT_GANTT_SEGMENT_NAME: () => DEFAULT_GANTT_SEGMENT_NAME,
  DEFAULT_TOOLTIP_MARGIN: () => DEFAULT_TOOLTIP_MARGIN,
  DISABLED_PATTERN_ANGLE: () => DISABLED_PATTERN_ANGLE,
  DISABLED_PATTERN_SIZE: () => DISABLED_PATTERN_SIZE,
  GANTT_CHART_V2_ARIA_LABEL: () => GANTT_CHART_V2_ARIA_LABEL,
  GANTT_LINE_SIZE_HEIGHT: () => GANTT_LINE_SIZE_HEIGHT,
  STRIPES_PATTERN_ID: () => STRIPES_PATTERN_ID,
  TIMESTAMP_WIDTH: () => TIMESTAMP_WIDTH
});
module.exports = __toCommonJS(defaults_exports);
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
const DEFAULT_GANTT_COLORS = {
  background: import_colors.default.Border.Neutral.Default,
  default: "blue"
};
const STRIPES_PATTERN_ID = "strato-gantt-chart-stripes-pattern";
const DISABLED_PATTERN_SIZE = 2;
const DISABLED_PATTERN_ANGLE = 45;
const DEFAULT_BAR_OPACITY = 0.85;
const DEFAULT_TOOLTIP_MARGIN = 40;
const DEFAULT_GANTT_LINE_SIZE = "default";
const DEFAULT_GANTT_SEGMENT_NAME = "Segment";
const GANTT_LINE_SIZE_HEIGHT = {
  default: 16,
  //px
  comfortable: 20,
  //px
  condensed: 12
  //px
};
const TIMESTAMP_WIDTH = 3;
const GANTT_CHART_V2_ARIA_LABEL = "Gantt Chart";
