// @dynatrace/strato-components/esm/charts/tree-map/components/TreeMapLegend/TreeMapRangeLegend.js
import { jsx } from "react/jsx-runtime";
import { useChartLayoutOrientation } from "../../../core/components/chart-layout/hooks/useChartLayoutOrientation.js";
import { getLegendOrientation } from "../../../core/components/chart-layout/utils/orientation.js";
import { RangeLegend } from "../../../core/components/legend/range/RangeLegend.js";
import { computeRangeLegendFormatter } from "../../../core/utils/compute-range-legend-formatter.js";
import { useFormatter } from "../../context/FormatterContext.js";
import { useTreeMapInternalData } from "../../context/TreeMapInternalDataContext.js";
import { useColorMap } from "../../store/selectors.js";
const TreeMapRangeLegend = () => {
  const legendColorMap = useColorMap();
  const { orientation: layoutOrientation } = useChartLayoutOrientation();
  const legendOrientation = getLegendOrientation(layoutOrientation);
  const { unit } = useTreeMapInternalData();
  const formatter = useFormatter();
  const computedFormatter = computeRangeLegendFormatter(formatter, unit);
  const legendItemKeys = [...legendColorMap.keys()];
  const items = legendItemKeys.map((key) => ({
    from: key.from,
    to: key.to,
    color: legendColorMap.get(key)?.color?.Default
  }));
  return /* @__PURE__ */ jsx(
    RangeLegend,
    {
      items,
      orientation: legendOrientation,
      formatter: computedFormatter
    }
  );
};
export {
  TreeMapRangeLegend
};
