// @dynatrace/strato-components/filters/timeframe-selector/utils/formatting.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var formatting_exports = {};
__export(formatting_exports, {
  formatTimeframeDuration: () => formatTimeframeDuration
});
module.exports = __toCommonJS(formatting_exports);
var import_date_fns = require("date-fns");
var import_timeframe_calculations = require("./timeframe-calculations.js");
var import_timeframeSelectorMessages = require("../constants/timeframeSelectorMessages.js");
function formatTimeframeDuration(from, to, intl) {
  if (from === null || to === null) {
    return "";
  }
  if (from.type === "iso8601" || to.type === "iso8601") {
    const difference2 = (0, import_date_fns.intervalToDuration)({
      start: from.absoluteDate,
      end: to.absoluteDate
    });
    const diffInMilliseconds = (0, import_date_fns.differenceInMilliseconds)(to.absoluteDate, from.absoluteDate) % 1e3;
    return intl.formatMessage(import_timeframeSelectorMessages.TIMEFRAME_SELECTOR_MESSAGES.absoluteDateDuration, {
      years: Math.abs(difference2.years ?? 0),
      months: Math.abs(difference2.months ?? 0),
      days: Math.abs(difference2.days ?? 0),
      hours: Math.abs(difference2.hours ?? 0),
      minutes: Math.abs(difference2.minutes ?? 0),
      seconds: Math.abs(difference2.seconds ?? 0),
      milliseconds: Math.abs(diffInMilliseconds ?? 0)
    }).trim();
  }
  const difference = (0, import_timeframe_calculations.getDifference)(from, to);
  const offsetUnit = (0, import_timeframe_calculations.getSmallestCommonUnit)(from, to);
  const timeExpression = import_timeframeSelectorMessages.TIMEFRAME_SELECTOR_MESSAGES[`${offsetUnit}-duration`];
  return intl.formatMessage(timeExpression, {
    offset: Math.abs(difference === 0 ? 1 : difference)
  }).trim();
}
