// @dynatrace/strato-components/filters/timeframe-selector/utils/presets-helpers.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var presets_helpers_exports = {};
__export(presets_helpers_exports, {
  addRecentTimeframePreset: () => addRecentTimeframePreset,
  getTimeframePresets: () => getTimeframePresets,
  isPresetSelected: () => isPresetSelected,
  isTimeframe: () => isTimeframe,
  isTimeframePreset: () => isTimeframePreset,
  isUserTimeframePresets: () => isUserTimeframePresets,
  saveRecentTimeframe: () => saveRecentTimeframe
});
module.exports = __toCommonJS(presets_helpers_exports);
var import_lodash_es = require("lodash");
var import_TimeframePresetsClient = require("./TimeframePresetsClient.js");
var import_logging = require("../../../core/utils/logging.js");
var import_timeParser = require("../../../forms/forms-core/time/timeParser.js");
function isTimeframe(timeframe) {
  if (typeof timeframe !== "object") {
    return false;
  }
  return typeof timeframe.from === "string" && typeof timeframe.to === "string";
}
function isTimeframePreset(timeframePreset) {
  if (typeof timeframePreset !== "object") {
    return false;
  }
  const hasValidId = timeframePreset.id !== void 0 && typeof timeframePreset.id === "string";
  const hasValidTimeframe = timeframePreset.timeframe && isTimeframe(timeframePreset.timeframe);
  return hasValidId && hasValidTimeframe;
}
function isUserTimeframePresets(userTimeframePreset) {
  if (typeof userTimeframePreset !== "object" || userTimeframePreset === null) {
    return false;
  }
  const hasValidRecentPresets = (0, import_lodash_es.isArray)(userTimeframePreset.recentPresets) && userTimeframePreset.recentPresets.every(
    (recentPreset) => isTimeframePreset(recentPreset)
  );
  return hasValidRecentPresets;
}
const RECENT_TIMEFRAME_SERVICE_UNAVAILABLE = "TimeframeSelector recent timeframe service is not available.";
const SAVE_RECENT_TIMEFRAME_ERROR = "TimeframeSelector failed to save recent timeframe.";
const GET_RECENT_TIMEFRAME_ERROR = "TimeframeSelector failed to get recent timeframes.";
async function addRecentTimeframePreset(timeframe) {
  const dtRuntimeTimeframes = globalThis.dtRuntime?.timeframes;
  if (!dtRuntimeTimeframes?.addRecentTimeframePreset) {
    (0, import_logging.devLog)(RECENT_TIMEFRAME_SERVICE_UNAVAILABLE);
    return { recentPresets: [] };
  }
  return dtRuntimeTimeframes.addRecentTimeframePreset({ timeframe }).then((presets) => {
    if (presets && isUserTimeframePresets(presets)) {
      return presets;
    }
    return { recentPresets: [] };
  }).catch(() => {
    (0, import_logging.devLog)(SAVE_RECENT_TIMEFRAME_ERROR);
    return { recentPresets: [] };
  });
}
function saveRecentTimeframe(timeframe) {
  void addRecentTimeframePreset({
    from: timeframe.from.value,
    to: timeframe.to.value
  }).then((presets) => {
    import_TimeframePresetsClient.timeframeQueryClient.setQueryData(import_TimeframePresetsClient.TIMEFRAME_PRESETS_QUERY_KEY, presets);
    return import_TimeframePresetsClient.timeframeQueryClient.refetchQueries({
      queryKey: import_TimeframePresetsClient.TIMEFRAME_PRESETS_QUERY_KEY,
      type: "all"
    });
  }).catch(() => {
  });
}
async function getTimeframePresets() {
  const dtRuntimeTimeframes = globalThis.dtRuntime?.timeframes;
  if (!dtRuntimeTimeframes?.getTimeframePresets) {
    (0, import_logging.devLog)(RECENT_TIMEFRAME_SERVICE_UNAVAILABLE);
    return { recentPresets: [] };
  }
  return dtRuntimeTimeframes.getTimeframePresets().then((presets) => {
    if (presets && isUserTimeframePresets(presets)) {
      return presets;
    }
    return { recentPresets: [] };
  }).catch(() => {
    (0, import_logging.devLog)(GET_RECENT_TIMEFRAME_ERROR);
    return { recentPresets: [] };
  });
}
function isPresetSelected(value, details, isRecent, precision) {
  if (isRecent) {
    return false;
  }
  const selectedFrom = details?.from?.value;
  const selectedTo = details?.to?.value;
  const { from: presetFrom, to: presetTo } = (0, import_timeParser.parseValue)(
    value,
    false,
    precision
  );
  return selectedFrom !== void 0 && selectedTo !== void 0 && selectedFrom === presetFrom?.value && selectedTo === presetTo?.value;
}
