// @dynatrace/strato-components/filters/timeframe-selector/utils/stepper-helpers.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var stepper_helpers_exports = {};
__export(stepper_helpers_exports, {
  formatStepperValues: () => formatStepperValues,
  getTimeShift: () => getTimeShift
});
module.exports = __toCommonJS(stepper_helpers_exports);
var import_tz = require("@date-fns/tz");
var import_date_fns = require("date-fns");
var import_units = require("@dynatrace-sdk/units");
var import_user_preferences = require("@dynatrace-sdk/user-preferences");
var import_formatting = require("./formatting.js");
var import_timeframe_calculations = require("./timeframe-calculations.js");
var import_timeframe_selection = require("./timeframe-selection.js");
var import_timeParser = require("../../../forms/forms-core/time/timeParser.js");
var import_timeframeSelectorMessages = require("../constants/timeframeSelectorMessages.js");
function formatStepperValues(from, to, intl, direction, selectedOverCalendar) {
  const adjustedTo = selectedOverCalendar && to ? (0, import_timeParser.getTimeValue)(
    (0, import_date_fns.startOfHour)(
      (0, import_date_fns.addHours)(new import_tz.TZDateMini(to.absoluteDate, (0, import_user_preferences.getTimezone)()), 1)
    ).toISOString()
  ) ?? null : to;
  const duration = (0, import_formatting.formatTimeframeDuration)(from, adjustedTo, intl);
  const prefix = intl.formatMessage(import_timeframeSelectorMessages.TIMEFRAME_SELECTOR_MESSAGES.prefix);
  const suffix = intl.formatMessage(import_timeframeSelectorMessages.TIMEFRAME_SELECTOR_MESSAGES.suffix, {
    suffix: direction === "backward" ? "back" : "forward"
  });
  return `${prefix}${duration}${suffix}`.trim();
}
const adjustDate = (date, difference, direction, selectedOverCalendar) => {
  if (selectedOverCalendar) {
    const daysDifference = Math.round(-difference / (1e3 * 60 * 60 * 24));
    date = (0, import_date_fns.addDays)(
      date,
      direction === "backward" ? -daysDifference : daysDifference
    );
    return date.toISOString();
  }
  date.setTime(
    direction === "backward" ? date.getTime() - Math.abs(difference) : date.getTime() + Math.abs(difference)
  );
  return date.toISOString();
};
const convertOffsetToExpression = (offsetAmount, diff, offsetUnit) => {
  const totalOffset = diff + offsetAmount;
  if (totalOffset > 0) {
    return `now()+${totalOffset}${offsetUnit}`;
  } else if (totalOffset < 0) {
    return `now()${totalOffset}${offsetUnit}`;
  } else {
    return "now()";
  }
};
const clampToMin = (min, max, from, to, precision) => {
  const relativeNow = import_tz.TZDateMini.tz((0, import_user_preferences.getTimezone)()).getTime();
  if (!min) {
    return { from, to };
  }
  const isFromBelowMin = (0, import_timeframe_selection.isSmallerThan)(from.absoluteDate, min.absoluteDate);
  const isToBelowMin = (0, import_timeframe_selection.isSmallerThan)(to.absoluteDate, min.absoluteDate);
  const difference = (0, import_date_fns.intervalToDuration)({
    start: from.absoluteDate,
    end: to.absoluteDate
  });
  if (isFromBelowMin && !isToBelowMin) {
    return {
      from: min,
      to
    };
  }
  const _timezonedMin = new import_tz.TZDateMini(min.absoluteDate, (0, import_user_preferences.getTimezone)());
  const adjustedTimeISO = new import_tz.TZDateMini(
    (0, import_date_fns.add)(_timezonedMin, difference),
    (0, import_user_preferences.getTimezone)()
  ).toISOString();
  let timeValueAdjusted = (0, import_units.parseTimeAsTimeValue)(
    adjustedTimeISO,
    relativeNow,
    precision
  );
  if (max && (0, import_timeframe_selection.isBiggerThan)(timeValueAdjusted?.absoluteDate, max.absoluteDate)) {
    timeValueAdjusted = max;
  }
  if (isFromBelowMin && isToBelowMin) {
    return {
      from: min,
      to: timeValueAdjusted ?? min
    };
  }
  if (!isFromBelowMin && isToBelowMin) {
    return {
      from,
      to: timeValueAdjusted ?? from
    };
  }
  return { from, to };
};
const clampToMax = (min, max, from, to, precision) => {
  const relativeNow = import_tz.TZDateMini.tz((0, import_user_preferences.getTimezone)()).getTime();
  if (!max) {
    return { from, to };
  }
  const isFromAboveMax = (0, import_timeframe_selection.isBiggerThan)(from.absoluteDate, max.absoluteDate);
  const isToAboveMax = (0, import_timeframe_selection.isBiggerThan)(to.absoluteDate, max.absoluteDate);
  const difference = (0, import_date_fns.intervalToDuration)({
    start: from.absoluteDate,
    end: to.absoluteDate
  });
  if (isFromAboveMax && isToAboveMax) {
    const _timezonedMax = new import_tz.TZDateMini(max.absoluteDate, (0, import_user_preferences.getTimezone)());
    const adjustedFrom = new import_tz.TZDateMini(
      (0, import_date_fns.sub)(_timezonedMax, difference),
      (0, import_user_preferences.getTimezone)()
    ).toISOString();
    if (min && (0, import_timeframe_selection.isSmallerThan)(adjustedFrom, min.absoluteDate)) {
      return { from: min, to: max };
    }
    return {
      from: (0, import_units.parseTimeAsTimeValue)(adjustedFrom, relativeNow, precision) ?? max,
      to: max
    };
  }
  if (!isFromAboveMax && isToAboveMax) {
    return {
      from,
      to: max
    };
  }
  if (isFromAboveMax && !isToAboveMax) {
    const _timezonedFrom = new import_tz.TZDateMini(from.absoluteDate, (0, import_user_preferences.getTimezone)());
    let adjustedTo = new import_tz.TZDateMini(
      (0, import_date_fns.sub)(_timezonedFrom, difference),
      (0, import_user_preferences.getTimezone)()
    ).toISOString();
    if (min && (0, import_timeframe_selection.isSmallerThan)(adjustedTo, min.absoluteDate)) {
      adjustedTo = min.absoluteDate;
    }
    return {
      from,
      to: (0, import_units.parseTimeAsTimeValue)(adjustedTo, relativeNow, precision) ?? from
    };
  }
  return { from, to };
};
const clampToBoundaries = (min, max, from, to, precision) => {
  const relativeNow = import_tz.TZDateMini.tz((0, import_user_preferences.getTimezone)()).getTime();
  const parsedMin = (0, import_units.parseTimeAsTimeValue)(min, relativeNow, precision);
  const parsedMax = (0, import_units.parseTimeAsTimeValue)(max, relativeNow, precision);
  let result = { from, to };
  result = clampToMin(parsedMin, parsedMax, result.from, result.to, precision);
  result = clampToMax(parsedMin, parsedMax, result.from, result.to, precision);
  return result;
};
const handleAbsolute = (originalValue, min, max, direction, selectedOverCalendar, precision) => {
  const timezone = (0, import_user_preferences.getTimezone)();
  const from = new import_tz.TZDateMini(originalValue.from.absoluteDate, timezone);
  const to = new import_tz.TZDateMini(originalValue.to.absoluteDate, timezone);
  const difference = from.getTime() - to.getTime();
  const adjustedFrom = (0, import_timeParser.removePrecision)(
    adjustDate(from, difference, direction, selectedOverCalendar),
    precision
  );
  const adjustedTo = (0, import_timeParser.removePrecision)(
    adjustDate(to, difference, direction, selectedOverCalendar),
    precision
  );
  const newTimeframe = {
    from: {
      type: "iso8601",
      absoluteDate: adjustedFrom,
      value: adjustedFrom
    },
    to: {
      type: "iso8601",
      absoluteDate: adjustedTo,
      value: adjustedTo
    }
  };
  return clampToBoundaries(
    min,
    max,
    newTimeframe.from,
    newTimeframe.to,
    precision
  ) ?? originalValue;
};
const handleExpression = (value, min, max, direction, precision) => {
  const relativeNow = import_tz.TZDateMini.tz((0, import_user_preferences.getTimezone)()).getTime();
  const fromDetails = (0, import_units._parseTimeExpression)(value.from.value);
  const toDetails = (0, import_units._parseTimeExpression)(value.to.value);
  const offsetUnit = (0, import_timeframe_calculations.getSmallestCommonUnit)(value.from, value.to);
  const fromOffset = offsetUnit === fromDetails?.offsetUnit ? fromDetails.offsetAmount : (0, import_timeframe_calculations.getTimeDifferenceInUnit)(fromDetails, relativeNow, offsetUnit);
  const toOffset = offsetUnit === toDetails?.offsetUnit ? toDetails.offsetAmount : (0, import_timeframe_calculations.getTimeDifferenceInUnit)(toDetails, relativeNow, offsetUnit);
  if (fromOffset === void 0 || toOffset === void 0) {
    return void 0;
  }
  const difference = Math.abs(fromOffset - toOffset);
  const newFromValue = convertOffsetToExpression(
    fromOffset,
    difference * -1,
    offsetUnit
  );
  const newToValue = convertOffsetToExpression(
    toOffset,
    difference,
    offsetUnit
  );
  const backwardFromParsedTime = (0, import_units.parseTimeAsTimeValue)(
    newFromValue,
    relativeNow,
    precision
  );
  const forwardToParsedTime = (0, import_units.parseTimeAsTimeValue)(
    newToValue,
    relativeNow,
    precision
  );
  if (!backwardFromParsedTime || !forwardToParsedTime) {
    return void 0;
  }
  const currentBackwardFromParsedTime = (0, import_units.parseTimeAsTimeValue)(
    convertOffsetToExpression(0, fromOffset, offsetUnit),
    relativeNow,
    precision
  );
  const currentForwardToParsedTime = (0, import_units.parseTimeAsTimeValue)(
    convertOffsetToExpression(0, toOffset, offsetUnit),
    relativeNow,
    precision
  );
  if (!currentBackwardFromParsedTime || !currentForwardToParsedTime) {
    return void 0;
  }
  const newTimeframe = direction === "backward" ? {
    from: backwardFromParsedTime,
    to: currentBackwardFromParsedTime
  } : {
    from: currentForwardToParsedTime,
    to: forwardToParsedTime
  };
  return clampToBoundaries(
    min,
    max,
    newTimeframe.from,
    newTimeframe.to,
    precision
  ) ?? value;
};
function handleRoundedTimeframe(value, min, max, direction, precision) {
  const relativeNow = import_tz.TZDateMini.tz((0, import_user_preferences.getTimezone)()).getTime();
  const fromDetails = (0, import_units._parseTimeExpression)(value.from.value);
  const toDetails = (0, import_units._parseTimeExpression)(value.to.value);
  const offsetUnit = (0, import_timeframe_calculations.getSmallestCommonUnit)(value.from, value.to);
  const fromOffset = (
    // same offsetUnit or now() or now()@d.
    offsetUnit === fromDetails?.offsetUnit || fromDetails?.offsetAmount === 0 ? fromDetails.offsetAmount : (0, import_timeframe_calculations.getTimeDifferenceInUnit)(fromDetails, relativeNow, offsetUnit)
  );
  const toOffset = (
    // same offsetUnit or now() or now()@d.
    offsetUnit === toDetails?.offsetUnit || toDetails?.offsetAmount === 0 ? toDetails.offsetAmount : (0, import_timeframe_calculations.getTimeDifferenceInUnit)(toDetails, relativeNow, offsetUnit)
  );
  let backwardFromValue = "";
  let forwardToValue = "";
  if (!fromDetails || !toDetails) {
    return void 0;
  }
  const roundDownUnit = fromDetails.roundDownUnit || toDetails.roundDownUnit;
  if (fromOffset === 0 && toOffset === 0) {
    backwardFromValue = `-1${offsetUnit}@${roundDownUnit}`;
    forwardToValue = `+1${offsetUnit}@${roundDownUnit}`;
  } else {
    const offsetDifference = Math.abs(fromOffset - toOffset);
    backwardFromValue = value.from.value === "now()" && offsetDifference === 1 ? (
      // Shift backward from now to the rounded down unit (e.g., @d).
      `@${roundDownUnit}`
    ) : (
      // Otherwise, if 'from' and 'to' are roundDownExpressions (e.g., -40d@M -> -20d@M).
      convertOffsetToExpression(
        offsetDifference * -1,
        fromOffset,
        offsetUnit
      ) + `@${roundDownUnit}`
    );
    forwardToValue = convertOffsetToExpression(offsetDifference, toOffset, offsetUnit) + `@${roundDownUnit}`;
  }
  const backwardFromParsedTime = (0, import_units.parseTimeAsTimeValue)(
    backwardFromValue,
    relativeNow,
    precision
  );
  const forwardToParsedTime = (0, import_units.parseTimeAsTimeValue)(
    forwardToValue,
    relativeNow,
    precision
  );
  if (!backwardFromParsedTime || !forwardToParsedTime) {
    return void 0;
  }
  const backwardFromCandidate = (
    // if the other unit is smaller than this unit convert it to that unit
    fromDetails?.offsetUnit && offsetUnit !== fromDetails?.offsetUnit ? (
      // Convert the offset to the smaller unit
      convertOffsetToExpression(0, fromOffset, offsetUnit) + `@${roundDownUnit}`
    ) : value.from.value
  );
  const forwardToCandidate = (
    // if the other unit is smaller than this unit convert it to that unit
    toDetails?.offsetUnit && offsetUnit !== toDetails?.offsetUnit ? (
      // Convert the offset to the smaller unit
      convertOffsetToExpression(0, toOffset, offsetUnit) + `@${roundDownUnit}`
    ) : value.to.value
  );
  const currentBackwardFromParsedTime = (0, import_units.parseTimeAsTimeValue)(
    backwardFromCandidate,
    relativeNow,
    precision
  );
  const currentForwardToParsedTime = (0, import_units.parseTimeAsTimeValue)(
    forwardToCandidate,
    relativeNow,
    precision
  );
  if (!currentBackwardFromParsedTime || !currentForwardToParsedTime) {
    return void 0;
  }
  const newTimeframe = direction === "backward" ? {
    from: backwardFromParsedTime,
    to: currentBackwardFromParsedTime
  } : {
    from: currentForwardToParsedTime,
    to: forwardToParsedTime
  };
  return clampToBoundaries(
    min,
    max,
    newTimeframe.from,
    newTimeframe.to,
    precision
  ) ?? value;
}
function getTimeShift(value, config) {
  const { direction, min, max, selectedOverCalendar, precision } = config;
  const isISO8601 = (type) => type === "iso8601";
  const isExpression = (type) => type === "expression";
  if (!value?.from || !value?.to) {
    return void 0;
  }
  if (isISO8601(value.from.type) || isISO8601(value.to.type)) {
    return handleAbsolute(
      value,
      min,
      max,
      direction,
      selectedOverCalendar,
      precision
    );
  }
  if (isExpression(value.from.type) && isExpression(value.to.type)) {
    const newFrom = (0, import_units._parseTimeExpression)(value.from.value);
    const newTo = (0, import_units._parseTimeExpression)(value.to.value);
    if (!(newFrom?.roundDownUnit || newTo?.roundDownUnit)) {
      return handleExpression(value, min, max, direction, precision);
    }
    if (
      // can be a roundedTimeframe or now().
      (newFrom?.roundDownUnit || newFrom?.offsetAmount === 0) && (newTo?.roundDownUnit || newTo?.offsetAmount === 0) && // checks if from and to have different roundDownUnits.
      !(newFrom?.roundDownUnit && newTo?.roundDownUnit && newFrom?.roundDownUnit !== newTo?.roundDownUnit)
    ) {
      return handleRoundedTimeframe(value, min, max, direction, precision);
    }
    return handleAbsolute(value, min, max, direction, false, precision);
  }
  return void 0;
}
