// @dynatrace/strato-components/filters/timeframe-selector/utils/timeframe-selection.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var timeframe_selection_exports = {};
__export(timeframe_selection_exports, {
  dateIsMinMax: () => dateIsMinMax,
  isBiggerThan: () => isBiggerThan,
  isOutsideRange: () => isOutsideRange,
  isSmallerThan: () => isSmallerThan,
  isTimeframeOrFromToString: () => isTimeframeOrFromToString
});
module.exports = __toCommonJS(timeframe_selection_exports);
var import_tz = require("@date-fns/tz");
var import_units = require("@dynatrace-sdk/units");
var import_user_preferences = require("@dynatrace-sdk/user-preferences");
var import_timeParser = require("../../../forms/forms-core/time/timeParser.js");
function dateIsMinMax(date, min, max) {
  if (!min && !max || !date) {
    return false;
  }
  const timezone = (0, import_user_preferences.getTimezone)();
  const relativeNow = import_tz.TZDateMini.tz(timezone).getTime();
  const isoDate = (0, import_units.parseTimeAsTimeValue)(date.value, relativeNow)?.absoluteDate;
  const tempDate = isoDate ? new import_tz.TZDateMini(isoDate, timezone) : null;
  if (!tempDate) {
    return false;
  }
  const minIsoDate = (0, import_units.parseTimeAsTimeValue)(min, relativeNow)?.absoluteDate;
  const maxIsoDate = (0, import_units.parseTimeAsTimeValue)(max, relativeNow)?.absoluteDate;
  const minDate = minIsoDate ? new import_tz.TZDateMini(minIsoDate, timezone) : null;
  const maxDate = maxIsoDate ? new import_tz.TZDateMini(maxIsoDate, timezone) : null;
  if (minDate && minDate.getTime() === tempDate.getTime()) {
    return true;
  }
  if (maxDate && maxDate.getTime() === tempDate.getTime()) {
    return true;
  }
  return false;
}
function isBiggerThan(candidate, threshold) {
  if (!threshold || !candidate) {
    return false;
  }
  const relativeNow = import_tz.TZDateMini.tz((0, import_user_preferences.getTimezone)()).getTime();
  const maxIsoDate = (0, import_units.parseTimeAsTimeValue)(threshold, relativeNow)?.absoluteDate;
  const valueIsoDate = (0, import_units.parseTimeAsTimeValue)(
    candidate,
    relativeNow
  )?.absoluteDate;
  if (!maxIsoDate || !valueIsoDate) {
    return false;
  }
  const maxDate = new import_tz.TZDateMini(maxIsoDate, (0, import_user_preferences.getTimezone)());
  const valueDate = new import_tz.TZDateMini(valueIsoDate, (0, import_user_preferences.getTimezone)());
  return valueDate.getTime() > maxDate.getTime();
}
function isSmallerThan(candidate, threshold) {
  if (!threshold || !candidate) {
    return false;
  }
  const relativeNow = import_tz.TZDateMini.tz((0, import_user_preferences.getTimezone)()).getTime();
  const minIsoDate = (0, import_units.parseTimeAsTimeValue)(threshold, relativeNow)?.absoluteDate;
  const valueIsoDate = (0, import_units.parseTimeAsTimeValue)(
    candidate,
    relativeNow
  )?.absoluteDate;
  if (!minIsoDate || !valueIsoDate) {
    return false;
  }
  const minDate = new import_tz.TZDateMini(minIsoDate, (0, import_user_preferences.getTimezone)());
  const valueDate = new import_tz.TZDateMini(valueIsoDate, (0, import_user_preferences.getTimezone)());
  return valueDate.getTime() < minDate.getTime();
}
function isOutsideRange(value, min, max) {
  if (!value || !min && !max) {
    return false;
  }
  const { valueFrom, valueTo } = (0, import_timeParser.getFromToValues)(value);
  if (min && (isSmallerThan(valueFrom, min) || isSmallerThan(valueTo, min))) {
    return true;
  }
  if (max && (isBiggerThan(valueFrom, max) || isBiggerThan(valueTo, max))) {
    return true;
  }
  return false;
}
function isTimeframeWithFromAndTo(date) {
  return typeof date?.from === "object" && typeof date?.to === "object" && date.from?.absoluteDate !== "" && date.to?.absoluteDate !== "";
}
function isTimeframeOrFromToString(date) {
  return typeof date?.from === "string" && typeof date?.to === "string" || isTimeframeWithFromAndTo(date);
}
