// @dynatrace/strato-components/filters/timeframe-selector/utils/TimeframePresetsClient.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TimeframePresetsClient_exports = {};
__export(TimeframePresetsClient_exports, {
  TIMEFRAME_PRESETS_QUERY_KEY: () => TIMEFRAME_PRESETS_QUERY_KEY,
  timeframeQueryClient: () => timeframeQueryClient
});
module.exports = __toCommonJS(TimeframePresetsClient_exports);
var import_react_query = require("@tanstack/react-query");
const TIMEFRAME_PRESETS_QUERY_KEY = ["getTimeframePresets"];
const timeframeQueryClient = new import_react_query.QueryClient({
  defaultOptions: {
    queries: {
      refetchOnWindowFocus: false,
      refetchOnMount: true,
      retry: false,
      staleTime: Infinity
    }
  }
});
