// @dynatrace/strato-components/filters/timeframe-selector/utils/timeframe-calculations.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var timeframe_calculations_exports = {};
__export(timeframe_calculations_exports, {
  getDifference: () => getDifference,
  getSmallestCommonUnit: () => getSmallestCommonUnit,
  getTimeDifferenceInUnit: () => getTimeDifferenceInUnit
});
module.exports = __toCommonJS(timeframe_calculations_exports);
var import_tz = require("@date-fns/tz");
var import_date_fns = require("date-fns");
var import_units = require("@dynatrace-sdk/units");
var import_user_preferences = require("@dynatrace-sdk/user-preferences");
const unitOrder = {
  y: 9,
  q: 8,
  M: 7,
  w: 6,
  w0: 6,
  w1: 6,
  w2: 6,
  w3: 6,
  w4: 6,
  w5: 6,
  w6: 6,
  d: 5,
  h: 4,
  m: 3,
  s: 2,
  ms: 1
};
function isSmallestCommonDenominatorUnit(left, right) {
  if (left === void 0) {
    return right ?? "ms";
  }
  if (right === void 0) {
    return left ?? "ms";
  }
  return unitOrder[left] < unitOrder[right] ? left : right;
}
function getSmallestCommonUnit(from, to) {
  const newFrom = (0, import_units._parseTimeExpression)(from?.value || "");
  const newTo = (0, import_units._parseTimeExpression)(to?.value || "");
  if (newFrom && newTo) {
    const unit = isSmallestCommonDenominatorUnit(
      // Use roundDownUnit only when there is no offsetUnit, such as in '@d -> now'.
      // For cases with both roundDownUnit and offsetUnit, like '+1M@d -> +2M@d',
      // rely solely on the offsetUnit to prevent unwanted conversions like '+1d@d'.
      newFrom.offsetUnit || newTo.offsetUnit || newFrom.roundDownUnit,
      newTo.offsetUnit || newFrom.offsetUnit || newTo.roundDownUnit
    );
    switch (unit) {
      case "w0":
      case "w1":
      case "w2":
      case "w3":
      case "w4":
      case "w5":
      case "w6":
        return "w";
      default:
        return unit;
    }
  }
  return "ms";
}
function getDifference(from, to) {
  const timezone = (0, import_user_preferences.getTimezone)();
  const relativeDate = import_tz.TZDateMini.tz(timezone).getTime();
  const newFrom = (0, import_units._parseTimeExpression)(from?.value || "");
  const newTo = (0, import_units._parseTimeExpression)(to?.value || "");
  if (newFrom && newTo) {
    const lowestUnit = getSmallestCommonUnit(from, to);
    return getTimeDifferenceInUnit(newFrom, relativeDate, lowestUnit) - getTimeDifferenceInUnit(newTo, relativeDate, lowestUnit);
  }
  if (from?.absoluteDate && to?.absoluteDate) {
    return Math.abs(
      new import_tz.TZDateMini(from.absoluteDate, timezone).getTime() - new import_tz.TZDateMini(to.absoluteDate, timezone).getTime()
    );
  }
  return 0;
}
function getTimeDifferenceInUnit(details, relativeDate, targetUnit) {
  if (!details) {
    return 0;
  }
  const newDate = new import_tz.TZDateMini(relativeDate, (0, import_user_preferences.getTimezone)());
  const expressionDate = (0, import_units._getDateFromExpression)(details, relativeDate);
  if (!expressionDate) {
    return 0;
  }
  switch (targetUnit) {
    case "ms":
      return (0, import_date_fns.differenceInMilliseconds)(expressionDate, newDate);
    case "s":
      return (0, import_date_fns.differenceInSeconds)(expressionDate, newDate);
    case "m":
      return (0, import_date_fns.differenceInMinutes)(expressionDate, newDate);
    case "h":
      return (0, import_date_fns.differenceInHours)(expressionDate, newDate);
    case "d":
      return (0, import_date_fns.differenceInCalendarDays)(expressionDate, newDate);
    case "w":
      return (0, import_date_fns.differenceInCalendarWeeks)(expressionDate, newDate);
    case "M":
      return (0, import_date_fns.differenceInCalendarMonths)(expressionDate, newDate);
    case "q":
      return (0, import_date_fns.differenceInCalendarQuarters)(expressionDate, newDate);
    case "y":
      return (0, import_date_fns.differenceInCalendarYears)(expressionDate, newDate);
    default:
      return (0, import_date_fns.differenceInMilliseconds)(expressionDate, newDate);
  }
}
