// @dynatrace/strato-components/core/dql/dql-common.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var dql_common_exports = {};
__export(dql_common_exports, {
  booleanLiteralPattern: () => booleanLiteralPattern,
  combinedDurationUnits: () => combinedDurationUnits,
  dqlCommands: () => dqlCommands,
  dqlUnitSuffixPattern: () => dqlUnitSuffixPattern,
  nullLiteralPattern: () => nullLiteralPattern,
  parameterNamePattern: () => parameterNamePattern,
  sortedDqlCommands: () => sortedDqlCommands
});
module.exports = __toCommonJS(dql_common_exports);
const dqlCommands = [
  "aggregate",
  "append",
  "data",
  "dedup",
  "describe",
  "expand",
  "experimentalFieldsSummary",
  "experimentalLoad",
  "experimentalSave",
  "experimentalTimeChart",
  "failIf",
  "fetch",
  "fields",
  "fieldsAdd",
  "fieldsFlatten",
  "fieldsKeep",
  "fieldsRemove",
  "fieldsRename",
  "fieldsSnapshot",
  "fieldsSummary",
  "filter",
  "filterOut",
  "group",
  "internalTimeseriesBins",
  "join",
  "joinNested",
  "limit",
  "load",
  "lookup",
  "makeTimeseries",
  "parse",
  "search",
  "smartscapeEdges",
  "smartscapeNodes",
  "sort",
  "summarize",
  "timeseries",
  "traverse"
];
const sortedDqlCommands = dqlCommands.toSorted(
  (a, b) => b.length - a.length
);
const parameterNamePattern = /\w+\s*(?=:)/;
const booleanLiteralPattern = /\b(?:true|false)\b/i;
const nullLiteralPattern = /\bnull\b/i;
const dqlDurationUnits = [
  "minutes",
  "minute",
  "hours",
  "hour",
  "mins",
  "secs",
  "hrs",
  "min",
  "sec",
  "ms",
  "ns",
  "us",
  "hr",
  "h",
  "s",
  "m"
];
const dqlCalendarDurationUnits = [
  "quarters",
  "quarter",
  "months",
  "weeks",
  "years",
  "month",
  "qtrs",
  "week",
  "year",
  "days",
  "day",
  "qtr",
  "yrs",
  "yr",
  "M",
  "w",
  "d",
  "q",
  "y"
];
const combinedDurationUnits = [
  ...dqlDurationUnits,
  ...dqlCalendarDurationUnits
];
const dqlUnitSuffixPattern = `(?:${combinedDurationUnits.join("|")})`;
