// @dynatrace/strato-components/charts/core/components/threshold-bar/hooks/useThresholdBarTooltip.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useThresholdBarTooltip_exports = {};
__export(useThresholdBarTooltip_exports, {
  useThresholdBarTooltip: () => useThresholdBarTooltip
});
module.exports = __toCommonJS(useThresholdBarTooltip_exports);
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_messages = require("../messages/messages.js");
var import_group_overlapping_markers = require("../utils/group-overlapping-markers.js");
const TOOLTIP_Y_OFFSET = 22;
function useThresholdBarTooltip(markers, onHover) {
  const groupedMarkers = (0, import_react.useMemo)(
    () => (0, import_group_overlapping_markers.groupOverlappingMarkers)(markers),
    [markers]
  );
  const intl = (0, import_react_intl.useIntl)();
  const [hoveredIndex, setHoveredIndex] = (0, import_react.useState)();
  const [tooltipPosition, setTooltipPosition] = (0, import_react.useState)();
  const [tooltipItems, setTooltipItems] = (0, import_react.useState)([]);
  (0, import_react.useEffect)(() => {
    if (hoveredIndex === void 0) {
      return;
    }
    const onScroll = () => setHoveredIndex(void 0);
    window.addEventListener("scroll", onScroll, { once: true, capture: true });
    return () => window.removeEventListener("scroll", onScroll, { capture: true });
  }, [hoveredIndex]);
  function handleMouseMove(event, index) {
    setHoveredIndex(index);
    const tooltipPosition2 = {
      x: event.clientX,
      y: event.clientY + TOOLTIP_Y_OFFSET
    };
    const marker = markers[index];
    const belongingGroup = groupedMarkers.find(
      (group) => group.some((item) => {
        return item.y1 === marker.y1 && item.y0 === marker.y0;
      })
    );
    if (belongingGroup) {
      const tooltipItems2 = belongingGroup.filter((item) => !item.isShadow).map((item, index2) => ({
        id: item.id,
        color: { Default: item.color },
        values: [item.value],
        name: item.label ?? intl.formatMessage(import_messages.THRESHOLD_MESSAGES.thresholdLabel)
      }));
      setTooltipPosition(tooltipPosition2);
      setTooltipItems(tooltipItems2);
      onHover?.(marker.id);
    }
  }
  function handleMouseLeave() {
    setHoveredIndex(void 0);
    setTooltipItems([]);
    onHover?.(void 0);
  }
  return {
    showTooltip: hoveredIndex !== void 0 && tooltipPosition,
    tooltipPosition,
    tooltipItems,
    handleMouseMove,
    handleMouseLeave
  };
}
