// @dynatrace/strato-components/charts/categorical-bar/components/interactions/tooltip/CategoricalBarChartTooltip.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CategoricalBarChartTooltip_exports = {};
__export(CategoricalBarChartTooltip_exports, {
  CategoricalBarChartTooltip: () => CategoricalBarChartTooltip
});
module.exports = __toCommonJS(CategoricalBarChartTooltip_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_categorical_tooltip = require("./categorical-tooltip.js");
var import_CategoricalBarChartTooltipWrapper = require("./CategoricalBarChartTooltipWrapper.js");
var import_ChartActionsMenu = require("../../../../core/components/series-action/ChartActionsMenu.js");
var import_useChartActionsMenu = require("../../../../core/components/series-action/hooks/useChartActionsMenu.js");
var import_useDefaultAction = require("../../../../core/components/series-action/hooks/useDefaultAction.js");
var import_InteractiveChartTooltip = require("../../../../core/components/tooltip/InteractiveChartTooltip/components/InteractiveChartTooltip.js");
var import_SeriesBasedContent = require("../../../../core/components/tooltip/InteractiveChartTooltip/contents/SeriesBasedContent.js");
var import_CategoricalTooltipHeader = require("../../../../core/components/tooltip/InteractiveChartTooltip/headers/CategoricalTooltipHeader.js");
var import_useSeriesBasedItemEstimatedSize = require("../../../../core/components/tooltip/InteractiveChartTooltip/hooks/useSeriesBasedItemEstimatedSize.js");
var import_useTooltipSeriesActions = require("../../../../core/components/tooltip/shared/hooks/useTooltipSeriesActions.js");
var import_safe_get = require("../../../../core/utils/safe-get.js");
var import_constants = require("../../../constants.js");
var import_selectors = require("../../../hooks/store/selectors.js");
var import_useCategoricalBarChartScales = require("../../../hooks/useCategoricalBarChartScales.js");
var import_useCategoricalBarChartTooltip = require("../../../hooks/useCategoricalBarChartTooltip.js");
var import_useCategoricalBarFilteredData = require("../../../hooks/useCategoricalBarFilteredData.js");
var import_useTruncationMode = require("../../../hooks/useTruncationMode.js");
var import_useSeriesActionClickHandler = require("../hooks/useSeriesActionClickHandler.js");
function CategoricalBarChartTooltip({
  closestDimensionPoint,
  hasHeader,
  formatter,
  unit
}) {
  const { visibleTooltip } = (0, import_selectors.useTooltipState)();
  const { variant } = (0, import_useCategoricalBarChartTooltip.useCategoricalBarChartTooltip)();
  const truncationMode = (0, import_useTruncationMode.useTruncationMode)();
  const {
    filteredCategories: categories,
    dimensionMap,
    categoryMap
  } = (0, import_useCategoricalBarFilteredData.useCategoricalBarFilteredData)();
  const {
    onLeave: handleMenuLeave,
    onScroll: handleMenuScroll,
    selectedItem,
    updateSelectedItem,
    actionsMenuRef,
    isScrolling
  } = (0, import_useChartActionsMenu.useChartActionsMenu)();
  const grouped = variant === import_constants.VARIANT_GROUPED;
  const closestCategory = (0, import_react.useMemo)(
    () => (0, import_categorical_tooltip.getClosestCategory)(categories, dimensionMap, closestDimensionPoint),
    [categories, dimensionMap, closestDimensionPoint]
  );
  const { valueScale } = (0, import_useCategoricalBarChartScales.useCategoricalBarChartScales)();
  const sortedTooltipItems = (0, import_react.useMemo)(
    () => (0, import_categorical_tooltip.getSortedTooltipItems)(
      dimensionMap,
      closestDimensionPoint,
      grouped,
      closestCategory,
      formatter,
      valueScale,
      unit
    ),
    [
      dimensionMap,
      closestDimensionPoint,
      grouped,
      closestCategory,
      formatter,
      valueScale,
      unit
    ]
  );
  const actionClickHandler = (0, import_useSeriesActionClickHandler.useSeriesActionClickHandler)(
    sortedTooltipItems.length > 1
  );
  const seriesActions = (0, import_useTooltipSeriesActions.useTooltipSeriesActions)(
    closestCategory?.dimensions,
    (dimension) => (0, import_safe_get.safeGet)(dimensionMap, dimension).id
  );
  const estimatedItemSize = (0, import_useSeriesBasedItemEstimatedSize.useSeriesBasedItemEstimatedSize)(
    sortedTooltipItems,
    "single-line"
  );
  const showTooltip = sortedTooltipItems.length && visibleTooltip;
  const headerItem = hasHeader ? closestDimensionPoint && sortedTooltipItems.find(
    (item) => item.id === closestDimensionPoint.datapoint.id
  ) : void 0;
  const bodyItems = headerItem ? sortedTooltipItems.filter((item) => item !== headerItem) : sortedTooltipItems;
  return showTooltip ? /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_CategoricalBarChartTooltipWrapper.CategoricalBarChartTooltipWrapper,
    {
      onMouseLeave: handleMenuLeave,
      onScrollContent: handleMenuScroll,
      estimatedItemSize,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_InteractiveChartTooltip.InteractiveChartTooltip.Header, { children: hasHeader && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_CategoricalTooltipHeader.CategoricalTooltipHeader,
          {
            name: (0, import_safe_get.safeGet)(categoryMap, closestCategory).name,
            item: headerItem
          }
        ) }),
        bodyItems && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_InteractiveChartTooltip.InteractiveChartTooltip.Body, { children: bodyItems.map((item, index) => () => {
          const isBodyItem = item !== headerItem;
          const isOpen = selectedItem === item.id;
          return isBodyItem && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_ChartActionsMenu.ChartActionsMenu,
            {
              isOpen,
              actions: isOpen ? seriesActions(item) : void 0,
              defaultActions: (0, import_useDefaultAction.getDefaultAction)(item),
              ref: actionsMenuRef,
              onActionClick: actionClickHandler,
              children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_SeriesBasedContent.SeriesBasedContent,
                {
                  item,
                  onMouseHover: () => updateSelectedItem(item.id),
                  showDivider: index !== bodyItems.length - 1,
                  truncationMode,
                  isScrolling
                }
              )
            },
            `${item.id}`
          );
        }) })
      ]
    }
  ) : null;
}
