// @dynatrace-sdk/client-filter-segment-management/cjs/index.js
/**
 * @license
 * Copyright 2023-2025 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/client/filter-segment-management/src/index.ts
var src_exports = {};
__export(src_exports, {
  ApiClientError: () => import_shared_errors4.ApiClientError,
  ApiGatewayError: () => import_shared_errors4.ApiGatewayError,
  ClientRequestError: () => import_shared_errors4.ClientRequestError,
  DetailedFilterSegmentAllowedOperationsItem: () => DetailedFilterSegmentAllowedOperationsItem,
  ErrorEnvelopeError: () => ErrorEnvelopeError,
  FilterSegmentAllowedOperationsItem: () => FilterSegmentAllowedOperationsItem,
  FilterSegmentsClient: () => FilterSegmentsClient,
  GetFilterSegmentQueryAddFieldsItem: () => GetFilterSegmentQueryAddFieldsItem,
  GetFilterSegmentsQueryAddFieldsItem: () => GetFilterSegmentsQueryAddFieldsItem,
  GetLeanFilterSegmentsQueryAddFieldsItem: () => GetLeanFilterSegmentsQueryAddFieldsItem,
  InvalidResponseError: () => import_shared_errors4.InvalidResponseError,
  LeanFilterSegmentAllowedOperationsItem: () => LeanFilterSegmentAllowedOperationsItem,
  filterSegmentsClient: () => filterSegmentsClient,
  isApiClientError: () => import_shared_errors4.isApiClientError,
  isApiGatewayError: () => import_shared_errors4.isApiGatewayError,
  isClientRequestError: () => import_shared_errors4.isClientRequestError,
  isErrorEnvelopeError: () => isErrorEnvelopeError,
  isInvalidResponseError: () => import_shared_errors4.isInvalidResponseError
});
module.exports = __toCommonJS(src_exports);

// packages/client/filter-segment-management/src/lib/apis/filter-segments-api.ts
var import_http_client = require("@dynatrace-sdk/http-client");
var import_shared_errors3 = require("@dynatrace-sdk/shared-errors");

// packages/client/filter-segment-management/src/lib/error-envelopes/api-gateway-errors-handler.ts
var import_shared_errors = require("@dynatrace-sdk/shared-errors");
async function apiGatewayErrorsHandler(response) {
  if (response.headers?.["dynatrace-response-source"] === "API Gateway") {
    const responseBody = await response.body("json");
    throw new import_shared_errors.ApiGatewayError(response, responseBody);
  }
}

// packages/client/filter-segment-management/src/lib/error-envelopes/error-envelope-error.ts
var import_shared_errors2 = require("@dynatrace-sdk/shared-errors");
var ErrorEnvelopeError = class extends import_shared_errors2.ClientRequestError {
  isErrorEnvelopeError = true;
};
function isErrorEnvelopeError(e) {
  return e?.isApiClientError === true && e?.isClientRequestError === true && e?.isErrorEnvelopeError === true && e instanceof Error;
}

// packages/client/filter-segment-management/src/lib/error-envelopes/get-error-message.ts
function serializeData(data) {
  try {
    return JSON.stringify(data);
  } catch (e) {
    return String(data);
  }
}
function getMessagesFromErrorDetails(details) {
  const messages = [];
  Object.entries(details).forEach(([name, data]) => {
    if (!data)
      return;
    const serializedData = serializeData(data);
    switch (name) {
      case "missingScopes":
        messages.push(`Missing scopes: ${serializedData}`);
        break;
      default:
        messages.push(`${name}: ${serializedData}`);
    }
  }, []);
  return messages;
}
function getErrorMessage(errorBody, defaultMessage) {
  const error = errorBody;
  const msg = error?.error?.message || defaultMessage;
  const details = error?.error?.details || {};
  return [msg, ...getMessagesFromErrorDetails(details)].join(". ");
}

// packages/client/filter-segment-management/src/lib/utils/transformations.ts
var defaultRuleConfig = {
  force: false,
  datetime: false
};
function mergeKeys(key1, key2) {
  if (!key1)
    return key2;
  if (!key2)
    return key1;
  return `${key1}.${key2}`;
}
function matchKey(key, rules, cache) {
  if (cache.has(key)) {
    return cache.get(key);
  }
  for (const rule of rules) {
    const pattern = Array.isArray(rule) ? rule[0] : rule;
    const config = Array.isArray(rule) ? rule[1] : defaultRuleConfig;
    if (key === pattern) {
      const result = [pattern, config];
      cache.set(key, result);
      return result;
    }
    let i = 0;
    let j = 0;
    const keyLen = key.length;
    const patternLen = pattern.length;
    let isMatch = true;
    while (i < keyLen && j < patternLen) {
      let keyPartEnd = key.indexOf(".", i);
      let patternPartEnd = pattern.indexOf(".", j);
      if (keyPartEnd === -1)
        keyPartEnd = keyLen;
      if (patternPartEnd === -1)
        patternPartEnd = patternLen;
      const keyPart = key.slice(i, keyPartEnd);
      const patternPart = pattern.slice(j, patternPartEnd);
      if (patternPart !== "*" && patternPart !== "**" && keyPart !== patternPart) {
        isMatch = false;
        break;
      }
      i = keyPartEnd + 1;
      j = patternPartEnd + 1;
    }
    if (isMatch) {
      if (j < patternLen) {
        const remaining = pattern.slice(j);
        if (remaining !== "**" && remaining !== "*") {
          isMatch = false;
        }
      }
      if (i < keyLen && pattern.slice(j - 1) !== "**") {
        isMatch = false;
      }
    }
    if (isMatch) {
      const result = [pattern, config];
      cache.set(key, result);
      return result;
    }
  }
  cache.set(key, null);
  return null;
}
var potentialMatchForRoot = (pattern, key) => pattern.startsWith(key) || pattern.startsWith("*") || pattern.startsWith("**");
var potentialMatch = (pattern, key) => pattern.startsWith(key) || pattern.includes("*") || pattern.includes("**");
function partialMatch(key, keyPatterns) {
  const keyPatternsLen = keyPatterns.length;
  if (!key.includes(".")) {
    for (let i = 0; i < keyPatternsLen; i++) {
      const pattern = keyPatterns[i];
      if (potentialMatchForRoot(pattern, key))
        return true;
    }
    return false;
  }
  let dotIndex = -1;
  let currentKeyEnd = 0;
  while ((dotIndex = key.indexOf(".", currentKeyEnd)) !== -1) {
    const currentKey = key.slice(0, dotIndex);
    const isRoot = !currentKey.includes(".");
    for (let i = 0; i < keyPatternsLen; i++) {
      const pattern = keyPatterns[i];
      if (isRoot) {
        if (potentialMatchForRoot(pattern, currentKey))
          return true;
      } else if (potentialMatch(pattern, currentKey))
        return true;
    }
    currentKeyEnd = dotIndex + 1;
  }
  for (let i = 0; i < keyPatternsLen; i++) {
    const pattern = keyPatterns[i];
    if (potentialMatch(pattern, key))
      return true;
  }
  return false;
}
var dateTimePattern = /^(\d{4}-\d{2}-\d{2})(T\d{2}:\d{2}:\d{2}(?:\.\d+)?(Z|[+-]\d{2}:\d{2})?)?$/i;
function isDateLike(value) {
  if (typeof value === "string") {
    return dateTimePattern.test(value);
  }
  return false;
}
function transformValue(value, rules, direction) {
  const { force, datetime } = rules;
  if (direction === "to") {
    return force || isDateLike(value) ? new Date(value) : value;
  }
  if (value instanceof Date) {
    const iso = value.toISOString();
    return datetime ? iso : iso.split("T")[0];
  }
  return value;
}
function walk(obj, direction, rules, rulesCache, keyPatterns, currentKey = "") {
  if (rules.length <= 0) {
    return obj;
  }
  if (!partialMatch(currentKey, keyPatterns)) {
    return obj;
  }
  if (Array.isArray(obj)) {
    const match = matchKey(currentKey, rules, rulesCache);
    obj.forEach((item, idx) => {
      if (typeof item === "object" && item !== null) {
        walk(item, direction, rules, rulesCache, keyPatterns, currentKey);
      } else if (match) {
        obj[idx] = transformValue(obj[idx], match[1], direction);
      }
    });
  }
  if (typeof obj === "object" && obj !== null && !Array.isArray(obj)) {
    for (const key of Object.keys(obj)) {
      const mergedKey = mergeKeys(currentKey, key);
      const match = matchKey(mergedKey, rules, rulesCache);
      if (match) {
        obj[key] = transformValue(obj[key], match[1], direction);
      }
      walk(obj[key], direction, rules, rulesCache, keyPatterns, mergedKey);
    }
  }
  return obj;
}
function flattenKeyPatterns(keyPatterns) {
  return keyPatterns.map((k) => Array.isArray(k) ? k[0] : k);
}
function transform(direction, object, rules) {
  const flatKeyPatterns = flattenKeyPatterns(rules);
  const rulesCache = /* @__PURE__ */ new Map();
  return walk(object, direction, rules, rulesCache, flatKeyPatterns);
}
function transformRequest(object, rules) {
  return transform("from", object, rules);
}
function transformResponse(object, rules) {
  return transform("to", object, rules);
}

// packages/client/filter-segment-management/src/lib/utils/url-helpers.ts
var encodeQueryParam = (key, value) => {
  const encodedKey = encodeURIComponent(key);
  return `${encodedKey}=${encodeURIComponent(typeof value === "number" ? value : String(value))}`;
};
var addExplodedArrayQueryParam = (query, key) => {
  const arrayValue = query[key];
  return arrayValue.map((value) => encodeQueryParam(key, value)).join("&");
};
var addNonExplodedArrayQueryParams = (query, key) => {
  const encodedKey = encodeURIComponent(key);
  const encodedParamsList = query[key].map((value) => encodeURIComponent(typeof value === "number" ? value : String(value))).join(",");
  return `${encodedKey}=${encodedParamsList}`;
};
var addQueryParam = (query, key) => encodeQueryParam(key, query[key]);
var arrayQueryParams = (query, key, explode) => {
  return explode ? addExplodedArrayQueryParam(query, key) : addNonExplodedArrayQueryParams(query, key);
};
var toQueryString = (rawQuery, flags = {}) => {
  const query = rawQuery || {};
  const keys = Object.keys(query).filter((key) => typeof query[key] !== "undefined");
  const queryString = keys.map(
    (key) => Array.isArray(query[key]) ? arrayQueryParams(query, key, flags.explode?.hasOwnProperty(key) ? flags.explode[key] : true) : addQueryParam(query, key)
  ).join("&");
  return queryString ? `?${queryString}` : "";
};

// packages/client/filter-segment-management/src/lib/apis/filter-segments-api.ts
var FilterSegmentsClient = class {
  httpClient;
  shouldTransformDates = true;
  /**
   * @param {HttpClient} httpClientImplementation - You can provide custom http client as a parameter to constructor of a FilterSegmentsClient. Custom http client should implement {HttpClient} interface.
   * @param {ApiClientOptions} [options] - Optional configuration for SDK Client.
   * @example
   * const filterSegmentsClientNodeJs = new FilterSegmentsClient(yourCustomImplementation);
   */
  constructor(httpClientImplementation, options) {
    this.httpClient = httpClientImplementation;
    this.shouldTransformDates = options?.transformDates ?? true;
  }
  /**
   * Get all filter-segments.
   *
   * Returns all filter-segments.
   * If details like description are not needed then consider using the `:lean` endpoint.
   *
   * Required scopes:
   *
   * - `storage:filter-segments:read`
   *
   * @returns {Promise<FilterSegments>} Filter-segments successfully retrieved.
   *
   * @example <caption>Code example</caption>
   * import { filterSegmentsClient } from "@dynatrace-sdk/client-filter-segment-management";
   *
   * const data = await filterSegmentsClient.getFilterSegments();
   *
   * @throws {ErrorEnvelopeError} bad request | unauthorized | missing permissions
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async getFilterSegments(config = {}) {
    if (!config) {
      throw new import_shared_errors3.ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const query = toQueryString({ "add-fields": config.addFields });
    try {
      const response = await this.httpClient.send({
        url: `/platform/storage/filter-segments/v1/filter-segments${query}`,
        method: "GET",
        headers: {
          Accept: "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
      } catch (err) {
        throw new import_shared_errors3.InvalidResponseError(
          `FilterSegmentsClient.getFilterSegments:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if ((0, import_http_client.isHttpClientAbortError)(e) || (0, import_shared_errors3.isInvalidResponseError)(e)) {
        throw e;
      }
      if (!(0, import_http_client.isHttpClientResponseError)(e)) {
        throw new import_shared_errors3.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new ErrorEnvelopeError(`400`, response, errorBody, getErrorMessage(errorBody, "bad request"), e);
          } catch (err) {
            if (err instanceof ErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `FilterSegmentsClient.getFilterSegments:400`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 401: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new ErrorEnvelopeError(`401`, response, errorBody, getErrorMessage(errorBody, "unauthorized"), e);
          } catch (err) {
            if (err instanceof ErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `FilterSegmentsClient.getFilterSegments:401`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new ErrorEnvelopeError(
              `403`,
              response,
              errorBody,
              getErrorMessage(errorBody, "missing permissions"),
              e
            );
          } catch (err) {
            if (err instanceof ErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `FilterSegmentsClient.getFilterSegments:403`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        default: {
          try {
            const responseValue = await response.body("json");
            throw new import_shared_errors3.ClientRequestError(
              `${response.status}`,
              response,
              responseValue,
              getErrorMessage(
                responseValue,
                `Unexpected api response: code=${response.status} body="${responseValue}"`
              ),
              e
            );
          } catch (error) {
            if (error instanceof import_shared_errors3.ClientRequestError) {
              throw error;
            }
            const responseValue = await response.body("text").catch(() => "");
            throw new import_shared_errors3.InvalidResponseError(
              `${response.status}`,
              error,
              responseValue,
              "json",
              "Unexpected api response",
              response
            );
          }
        }
      }
    }
  }
  /**
   * Create a new filter-segment.
   *
   * Create a new segment and define data.
   *
   * Required scopes:
   *
   * - `storage:filter-segments:write`
   * - `storage:filter-segments:share` if `isPublic` is set to `true`
   *
   * ### New Filter-Segment
   *
   * Provide a name and an optional description for the filter-segment.
   *
   * #### Variables
   *
   * Create variables to apply in your filter-segment data filters
   *
   * #### Includes (Segment data)
   *
   * Include all data that should be accessible when applying the filter-segment
   *
   * #### IsPublic (Visibility)
   *
   * - `false`: The filter-segment is private and only visible to the owner
   * - `true`: The filter-segment is visible to everyone in the environment
   *
   * @returns {Promise<DetailedFilterSegment>} Filter-segment successfully created.
   *
   * @example <caption>Code example</caption>
   * import { filterSegmentsClient } from "@dynatrace-sdk/client-filter-segment-management";
   *
   * const data = await filterSegmentsClient.createFilterSegment(
   *   {
   *     body: {
   *       name: "dev_environment",
   *       description:
   *         "only includes data of the dev environment",
   *       variables: {
   *         type: "query",
   *         value: "fetch logs | limit 1",
   *       },
   *       isPublic: false,
   *       includes: [
   *         {
   *           filter:
   *             '{"type":"Group","range":{"from":0,"to":19},"logicalOperator":"AND","explicit":false,"children":[{"type":"Statement","range":{"from":0,"to":19},"key":{"type":"Key","textValue":"content","value":"content","range":{"from":0,"to":7}},"operator":{"type":"ComparisonOperator","textValue":"=","value":"=","range":{"from":8,"to":9}},"value":{"type":"String","textValue":"\\"[debug]\\"","value":"[debug]","range":{"from":10,"to":19},"isEscaped":true}}]}',
   *           dataObject: "logs",
   *         },
   *         {
   *           filter:
   *             '{"type":"Group","range":{"from":0,"to":18},"logicalOperator":"AND","explicit":false,"children":[{"type":"Statement","range":{"from":0,"to":18},"key":{"type":"Key","textValue":"content","value":"content","range":{"from":0,"to":7}},"operator":{"type":"ComparisonOperator","textValue":"=","value":"=","range":{"from":8,"to":9}},"value":{"type":"String","textValue":"\\"[info]\\"","value":"[info]","range":{"from":10,"to":18},"isEscaped":true}}]}',
   *           dataObject: "events",
   *         },
   *       ],
   *     },
   *   },
   * );
   *
   * @throws {ErrorEnvelopeError} Filter-segment Bad Request. | unauthorized | missing permissions
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async createFilterSegment(config) {
    if (!config) {
      throw new import_shared_errors3.ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const encodedBody = this.shouldTransformDates ? transformRequest(config.body, []) : config.body;
    try {
      const response = await this.httpClient.send({
        url: `/platform/storage/filter-segments/v1/filter-segments`,
        method: "POST",
        requestBodyType: "json",
        body: encodedBody,
        headers: {
          "Content-Type": "application/json",
          Accept: "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [201].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
      } catch (err) {
        throw new import_shared_errors3.InvalidResponseError(
          `FilterSegmentsClient.createFilterSegment:201`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if ((0, import_http_client.isHttpClientAbortError)(e) || (0, import_shared_errors3.isInvalidResponseError)(e)) {
        throw e;
      }
      if (!(0, import_http_client.isHttpClientResponseError)(e)) {
        throw new import_shared_errors3.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new ErrorEnvelopeError(
              `400`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Filter-segment Bad Request."),
              e
            );
          } catch (err) {
            if (err instanceof ErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `FilterSegmentsClient.createFilterSegment:400`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 401: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new ErrorEnvelopeError(`401`, response, errorBody, getErrorMessage(errorBody, "unauthorized"), e);
          } catch (err) {
            if (err instanceof ErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `FilterSegmentsClient.createFilterSegment:401`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new ErrorEnvelopeError(
              `403`,
              response,
              errorBody,
              getErrorMessage(errorBody, "missing permissions"),
              e
            );
          } catch (err) {
            if (err instanceof ErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `FilterSegmentsClient.createFilterSegment:403`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        default: {
          try {
            const responseValue = await response.body("json");
            throw new import_shared_errors3.ClientRequestError(
              `${response.status}`,
              response,
              responseValue,
              getErrorMessage(
                responseValue,
                `Unexpected api response: code=${response.status} body="${responseValue}"`
              ),
              e
            );
          } catch (error) {
            if (error instanceof import_shared_errors3.ClientRequestError) {
              throw error;
            }
            const responseValue = await response.body("text").catch(() => "");
            throw new import_shared_errors3.InvalidResponseError(
              `${response.status}`,
              error,
              responseValue,
              "json",
              "Unexpected api response",
              response
            );
          }
        }
      }
    }
  }
  /**
   * Get filter-segment-entity-model.
   *
   * Returns the filter-segment-entity-model.
   *
   * Required scopes:
   *
   * - `storage:filter-segments:read`
   *
   * @returns {Promise<FilterSegmentNamespaceDto>} Filter-segments-entity-model successfully retrieved.
   *
   * @example <caption>Code example</caption>
   * import { filterSegmentsClient } from "@dynatrace-sdk/client-filter-segment-management";
   *
   * const data =
   *   await filterSegmentsClient.getFilterSegmentsEntityModel();
   *
   * @throws {ErrorEnvelopeError} bad request | unauthorized | missing permissions
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async getFilterSegmentsEntityModel(config = {}) {
    try {
      const response = await this.httpClient.send({
        url: `/platform/storage/filter-segments/v1/filter-segments-entity-model`,
        method: "GET",
        headers: {
          Accept: "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
      } catch (err) {
        throw new import_shared_errors3.InvalidResponseError(
          `FilterSegmentsClient.getFilterSegmentsEntityModel:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if ((0, import_http_client.isHttpClientAbortError)(e) || (0, import_shared_errors3.isInvalidResponseError)(e)) {
        throw e;
      }
      if (!(0, import_http_client.isHttpClientResponseError)(e)) {
        throw new import_shared_errors3.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new ErrorEnvelopeError(`400`, response, errorBody, getErrorMessage(errorBody, "bad request"), e);
          } catch (err) {
            if (err instanceof ErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `FilterSegmentsClient.getFilterSegmentsEntityModel:400`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 401: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new ErrorEnvelopeError(`401`, response, errorBody, getErrorMessage(errorBody, "unauthorized"), e);
          } catch (err) {
            if (err instanceof ErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `FilterSegmentsClient.getFilterSegmentsEntityModel:401`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new ErrorEnvelopeError(
              `403`,
              response,
              errorBody,
              getErrorMessage(errorBody, "missing permissions"),
              e
            );
          } catch (err) {
            if (err instanceof ErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `FilterSegmentsClient.getFilterSegmentsEntityModel:403`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        default: {
          try {
            const responseValue = await response.body("json");
            throw new import_shared_errors3.ClientRequestError(
              `${response.status}`,
              response,
              responseValue,
              getErrorMessage(
                responseValue,
                `Unexpected api response: code=${response.status} body="${responseValue}"`
              ),
              e
            );
          } catch (error) {
            if (error instanceof import_shared_errors3.ClientRequestError) {
              throw error;
            }
            const responseValue = await response.body("text").catch(() => "");
            throw new import_shared_errors3.InvalidResponseError(
              `${response.status}`,
              error,
              responseValue,
              "json",
              "Unexpected api response",
              response
            );
          }
        }
      }
    }
  }
  /**
   * Get filter-segment by UID.
   *
   * Get filter-segment by its UID.
   *
   * Required scopes:
   *
   * - `storage:filter-segments:read`
   *
   * @returns {Promise<DetailedFilterSegment>} Filter-segment successfully retrieved.
   *
   * @example <caption>Code example</caption>
   * import { filterSegmentsClient } from "@dynatrace-sdk/client-filter-segment-management";
   *
   * const data = await filterSegmentsClient.getFilterSegment({
   *   filterSegmentUid: "D82a1jdA23a",
   * });
   *
   * @throws {ErrorEnvelopeError} bad request | unauthorized | missing permissions | Not found: the server cannot find the requested filter-segment.
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async getFilterSegment(config) {
    if (!config) {
      throw new import_shared_errors3.ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const query = toQueryString({ "add-fields": config.addFields });
    try {
      const response = await this.httpClient.send({
        url: `/platform/storage/filter-segments/v1/filter-segments/${config.filterSegmentUid}${query}`,
        method: "GET",
        headers: {
          Accept: "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
      } catch (err) {
        throw new import_shared_errors3.InvalidResponseError(
          `FilterSegmentsClient.getFilterSegment:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if ((0, import_http_client.isHttpClientAbortError)(e) || (0, import_shared_errors3.isInvalidResponseError)(e)) {
        throw e;
      }
      if (!(0, import_http_client.isHttpClientResponseError)(e)) {
        throw new import_shared_errors3.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new ErrorEnvelopeError(`400`, response, errorBody, getErrorMessage(errorBody, "bad request"), e);
          } catch (err) {
            if (err instanceof ErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `FilterSegmentsClient.getFilterSegment:400`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 401: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new ErrorEnvelopeError(`401`, response, errorBody, getErrorMessage(errorBody, "unauthorized"), e);
          } catch (err) {
            if (err instanceof ErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `FilterSegmentsClient.getFilterSegment:401`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new ErrorEnvelopeError(
              `403`,
              response,
              errorBody,
              getErrorMessage(errorBody, "missing permissions"),
              e
            );
          } catch (err) {
            if (err instanceof ErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `FilterSegmentsClient.getFilterSegment:403`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new ErrorEnvelopeError(
              `404`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Not found: the server cannot find the requested filter-segment."),
              e
            );
          } catch (err) {
            if (err instanceof ErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `FilterSegmentsClient.getFilterSegment:404`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        default: {
          try {
            const responseValue = await response.body("json");
            throw new import_shared_errors3.ClientRequestError(
              `${response.status}`,
              response,
              responseValue,
              getErrorMessage(
                responseValue,
                `Unexpected api response: code=${response.status} body="${responseValue}"`
              ),
              e
            );
          } catch (error) {
            if (error instanceof import_shared_errors3.ClientRequestError) {
              throw error;
            }
            const responseValue = await response.body("text").catch(() => "");
            throw new import_shared_errors3.InvalidResponseError(
              `${response.status}`,
              error,
              responseValue,
              "json",
              "Unexpected api response",
              response
            );
          }
        }
      }
    }
  }
  /**
   * Update a filter-segment. All fields will be overwritten (also undefined or null fields).
   *
   * Update one or more fields of a filter-segment.
   *
   * Required scopes:
   * - `storage:filter-segments:write`
   * - `storage:filter-segments:share` to change `isPublic` to `true`
   *
   * Following fields can be updated:
   *
   * - `name`
   * - `description`
   * - `isPublic`
   * - `variables`*
   * - `includes`*
   *
   * \* Values of the fields will be overridden or removed if not given.
   *
   * Provide all other fields with unchanged values.
   *
   * @example <caption>Code example</caption>
   * import { filterSegmentsClient } from "@dynatrace-sdk/client-filter-segment-management";
   *
   * const data = await filterSegmentsClient.updateFilterSegment(
   *   {
   *     filterSegmentUid: "D82a1jdA23a",
   *     optimisticLockingVersion: 1,
   *     body: {
   *       uid: "D82a1jdA23a",
   *       name: "dev_environment",
   *       description:
   *         "only includes data of the development environment",
   *       variables: {
   *         type: "query",
   *         value: "fetch logs | limit 1",
   *       },
   *       isPublic: false,
   *       owner: "john.doe",
   *       includes: [
   *         {
   *           filter:
   *             '{"type":"Group","range":{"from":0,"to":19},"logicalOperator":"AND","explicit":false,"children":[{"type":"Statement","range":{"from":0,"to":19},"key":{"type":"Key","textValue":"content","value":"content","range":{"from":0,"to":7}},"operator":{"type":"ComparisonOperator","textValue":"=","value":"=","range":{"from":8,"to":9}},"value":{"type":"String","textValue":"\\"[debug]\\"","value":"[debug]","range":{"from":10,"to":19},"isEscaped":true}}]}',
   *           dataObject: "logs",
   *         },
   *         {
   *           filter:
   *             '{"type":"Group","range":{"from":0,"to":18},"logicalOperator":"AND","explicit":false,"children":[{"type":"Statement","range":{"from":0,"to":18},"key":{"type":"Key","textValue":"content","value":"content","range":{"from":0,"to":7}},"operator":{"type":"ComparisonOperator","textValue":"=","value":"=","range":{"from":8,"to":9}},"value":{"type":"String","textValue":"\\"[info]\\"","value":"[info]","range":{"from":10,"to":18},"isEscaped":true}}]}',
   *           dataObject: "events",
   *         },
   *       ],
   *     },
   *   },
   * );
   *
   * @throws {ErrorEnvelopeError} bad request | unauthorized | missing permissions | Not found: the server cannot find the requested filter-segment.
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async updateFilterSegment(config) {
    if (!config) {
      throw new import_shared_errors3.ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const encodedBody = this.shouldTransformDates ? transformRequest(config.body, []) : config.body;
    const query = toQueryString({ "optimistic-locking-version": config.optimisticLockingVersion });
    try {
      await this.httpClient.send({
        url: `/platform/storage/filter-segments/v1/filter-segments/${config.filterSegmentUid}${query}`,
        method: "PUT",
        requestBodyType: "json",
        body: encodedBody,
        headers: {
          "Content-Type": "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      return;
    } catch (e) {
      if ((0, import_http_client.isHttpClientAbortError)(e)) {
        throw e;
      }
      if (!(0, import_http_client.isHttpClientResponseError)(e)) {
        throw new import_shared_errors3.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new ErrorEnvelopeError(`400`, response, errorBody, getErrorMessage(errorBody, "bad request"), e);
          } catch (err) {
            if (err instanceof ErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `FilterSegmentsClient.updateFilterSegment:400`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 401: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new ErrorEnvelopeError(`401`, response, errorBody, getErrorMessage(errorBody, "unauthorized"), e);
          } catch (err) {
            if (err instanceof ErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `FilterSegmentsClient.updateFilterSegment:401`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new ErrorEnvelopeError(
              `403`,
              response,
              errorBody,
              getErrorMessage(errorBody, "missing permissions"),
              e
            );
          } catch (err) {
            if (err instanceof ErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `FilterSegmentsClient.updateFilterSegment:403`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new ErrorEnvelopeError(
              `404`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Not found: the server cannot find the requested filter-segment."),
              e
            );
          } catch (err) {
            if (err instanceof ErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `FilterSegmentsClient.updateFilterSegment:404`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        default: {
          try {
            const responseValue = await response.body("json");
            throw new import_shared_errors3.ClientRequestError(
              `${response.status}`,
              response,
              responseValue,
              getErrorMessage(
                responseValue,
                `Unexpected api response: code=${response.status} body="${responseValue}"`
              ),
              e
            );
          } catch (error) {
            if (error instanceof import_shared_errors3.ClientRequestError) {
              throw error;
            }
            const responseValue = await response.body("text").catch(() => "");
            throw new import_shared_errors3.InvalidResponseError(
              `${response.status}`,
              error,
              responseValue,
              "json",
              "Unexpected api response",
              response
            );
          }
        }
      }
    }
  }
  /**
   * Delete a filter-segment.
   *
   * Remove the filter-segment from the environment for all users.
   *
   * Required scopes:
   * - `storage:filter-segments:delete`
   *
   * @example <caption>Code example</caption>
   * import { filterSegmentsClient } from "@dynatrace-sdk/client-filter-segment-management";
   *
   * const data = await filterSegmentsClient.deleteFilterSegment(
   *   { filterSegmentUid: "D82a1jdA23a" },
   * );
   *
   * @throws {ErrorEnvelopeError} bad request | unauthorized | missing permissions | Not found: the server cannot find the requested filter-segment.
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async deleteFilterSegment(config) {
    try {
      await this.httpClient.send({
        url: `/platform/storage/filter-segments/v1/filter-segments/${config.filterSegmentUid}`,
        method: "DELETE",
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [204].includes(status);
        }
      });
      return;
    } catch (e) {
      if ((0, import_http_client.isHttpClientAbortError)(e)) {
        throw e;
      }
      if (!(0, import_http_client.isHttpClientResponseError)(e)) {
        throw new import_shared_errors3.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new ErrorEnvelopeError(`400`, response, errorBody, getErrorMessage(errorBody, "bad request"), e);
          } catch (err) {
            if (err instanceof ErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `FilterSegmentsClient.deleteFilterSegment:400`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 401: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new ErrorEnvelopeError(`401`, response, errorBody, getErrorMessage(errorBody, "unauthorized"), e);
          } catch (err) {
            if (err instanceof ErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `FilterSegmentsClient.deleteFilterSegment:401`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new ErrorEnvelopeError(
              `403`,
              response,
              errorBody,
              getErrorMessage(errorBody, "missing permissions"),
              e
            );
          } catch (err) {
            if (err instanceof ErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `FilterSegmentsClient.deleteFilterSegment:403`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new ErrorEnvelopeError(
              `404`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Not found: the server cannot find the requested filter-segment."),
              e
            );
          } catch (err) {
            if (err instanceof ErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `FilterSegmentsClient.deleteFilterSegment:404`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        default: {
          try {
            const responseValue = await response.body("json");
            throw new import_shared_errors3.ClientRequestError(
              `${response.status}`,
              response,
              responseValue,
              getErrorMessage(
                responseValue,
                `Unexpected api response: code=${response.status} body="${responseValue}"`
              ),
              e
            );
          } catch (error) {
            if (error instanceof import_shared_errors3.ClientRequestError) {
              throw error;
            }
            const responseValue = await response.body("text").catch(() => "");
            throw new import_shared_errors3.InvalidResponseError(
              `${response.status}`,
              error,
              responseValue,
              "json",
              "Unexpected api response",
              response
            );
          }
        }
      }
    }
  }
  /**
   * Partially update a filter-segment.
   *
   * Update one or more fields of a filter-segment.
   *
   * Required scopes:
   * - `storage:filter-segments:write`
   * - `storage:filter-segments:share` to change `isPublic` to `true`
   *
   * Following fields can be updated:
   *
   * - `name`
   * - `description`
   * - `isPublic`
   * - `variables`*
   * - `includes`*
   *
   * \* If given the value will be overridden.
   *
   * Fields not given will be ignored.
   *
   * @example <caption>Code example</caption>
   * import { filterSegmentsClient } from "@dynatrace-sdk/client-filter-segment-management";
   *
   * const data =
   *   await filterSegmentsClient.partiallyUpdateFilterSegment({
   *     filterSegmentUid: "D82a1jdA23a",
   *     optimisticLockingVersion: 1,
   *     body: {
   *       description:
   *         "only includes data of the development environment",
   *     },
   *   });
   *
   * @throws {ErrorEnvelopeError} bad request | unauthorized | missing permissions | Not found: the server cannot find the requested filter-segment.
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async partiallyUpdateFilterSegment(config) {
    if (!config) {
      throw new import_shared_errors3.ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const encodedBody = this.shouldTransformDates ? transformRequest(config.body, []) : config.body;
    const query = toQueryString({ "optimistic-locking-version": config.optimisticLockingVersion });
    try {
      await this.httpClient.send({
        url: `/platform/storage/filter-segments/v1/filter-segments/${config.filterSegmentUid}${query}`,
        method: "PATCH",
        requestBodyType: "json",
        body: encodedBody,
        headers: {
          "Content-Type": "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      return;
    } catch (e) {
      if ((0, import_http_client.isHttpClientAbortError)(e)) {
        throw e;
      }
      if (!(0, import_http_client.isHttpClientResponseError)(e)) {
        throw new import_shared_errors3.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new ErrorEnvelopeError(`400`, response, errorBody, getErrorMessage(errorBody, "bad request"), e);
          } catch (err) {
            if (err instanceof ErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `FilterSegmentsClient.partiallyUpdateFilterSegment:400`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 401: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new ErrorEnvelopeError(`401`, response, errorBody, getErrorMessage(errorBody, "unauthorized"), e);
          } catch (err) {
            if (err instanceof ErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `FilterSegmentsClient.partiallyUpdateFilterSegment:401`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new ErrorEnvelopeError(
              `403`,
              response,
              errorBody,
              getErrorMessage(errorBody, "missing permissions"),
              e
            );
          } catch (err) {
            if (err instanceof ErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `FilterSegmentsClient.partiallyUpdateFilterSegment:403`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new ErrorEnvelopeError(
              `404`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Not found: the server cannot find the requested filter-segment."),
              e
            );
          } catch (err) {
            if (err instanceof ErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `FilterSegmentsClient.partiallyUpdateFilterSegment:404`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        default: {
          try {
            const responseValue = await response.body("json");
            throw new import_shared_errors3.ClientRequestError(
              `${response.status}`,
              response,
              responseValue,
              getErrorMessage(
                responseValue,
                `Unexpected api response: code=${response.status} body="${responseValue}"`
              ),
              e
            );
          } catch (error) {
            if (error instanceof import_shared_errors3.ClientRequestError) {
              throw error;
            }
            const responseValue = await response.body("text").catch(() => "");
            throw new import_shared_errors3.InvalidResponseError(
              `${response.status}`,
              error,
              responseValue,
              "json",
              "Unexpected api response",
              response
            );
          }
        }
      }
    }
  }
  /**
   * Get all filter-segments in a minimal/lean form.
   *
   * Returns all filter-segments in a minimal representational form. This endpoint saves resources
   * and as a result quicker response times can be expected over the more detailed get-all
   * endpoint.
   *
   * Required scopes:
   *
   * - `storage:filter-segments:read`
   *
   * @returns {Promise<LeanFilterSegments>} Filter-segments successfully retrieved.
   *
   * @example <caption>Code example</caption>
   * import { filterSegmentsClient } from "@dynatrace-sdk/client-filter-segment-management";
   *
   * const data =
   *   await filterSegmentsClient.getLeanFilterSegments();
   *
   * @throws {ErrorEnvelopeError} bad request | unauthorized | missing permissions
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async getLeanFilterSegments(config = {}) {
    if (!config) {
      throw new import_shared_errors3.ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const query = toQueryString({ "add-fields": config.addFields });
    try {
      const response = await this.httpClient.send({
        url: `/platform/storage/filter-segments/v1/filter-segments:lean${query}`,
        method: "GET",
        headers: {
          Accept: "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
      } catch (err) {
        throw new import_shared_errors3.InvalidResponseError(
          `FilterSegmentsClient.getLeanFilterSegments:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if ((0, import_http_client.isHttpClientAbortError)(e) || (0, import_shared_errors3.isInvalidResponseError)(e)) {
        throw e;
      }
      if (!(0, import_http_client.isHttpClientResponseError)(e)) {
        throw new import_shared_errors3.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new ErrorEnvelopeError(`400`, response, errorBody, getErrorMessage(errorBody, "bad request"), e);
          } catch (err) {
            if (err instanceof ErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `FilterSegmentsClient.getLeanFilterSegments:400`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 401: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new ErrorEnvelopeError(`401`, response, errorBody, getErrorMessage(errorBody, "unauthorized"), e);
          } catch (err) {
            if (err instanceof ErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `FilterSegmentsClient.getLeanFilterSegments:401`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new ErrorEnvelopeError(
              `403`,
              response,
              errorBody,
              getErrorMessage(errorBody, "missing permissions"),
              e
            );
          } catch (err) {
            if (err instanceof ErrorEnvelopeError) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `FilterSegmentsClient.getLeanFilterSegments:403`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        default: {
          try {
            const responseValue = await response.body("json");
            throw new import_shared_errors3.ClientRequestError(
              `${response.status}`,
              response,
              responseValue,
              getErrorMessage(
                responseValue,
                `Unexpected api response: code=${response.status} body="${responseValue}"`
              ),
              e
            );
          } catch (error) {
            if (error instanceof import_shared_errors3.ClientRequestError) {
              throw error;
            }
            const responseValue = await response.body("text").catch(() => "");
            throw new import_shared_errors3.InvalidResponseError(
              `${response.status}`,
              error,
              responseValue,
              "json",
              "Unexpected api response",
              response
            );
          }
        }
      }
    }
  }
};
var filterSegmentsClient = /* @__PURE__ */ new FilterSegmentsClient(import_http_client.httpClient);

// packages/client/filter-segment-management/src/lib/error-envelopes/index.ts
var import_shared_errors4 = require("@dynatrace-sdk/shared-errors");

// packages/client/filter-segment-management/src/lib/models/detailed-filter-segment-allowed-operations-item.ts
var DetailedFilterSegmentAllowedOperationsItem = /* @__PURE__ */ ((DetailedFilterSegmentAllowedOperationsItem2) => {
  DetailedFilterSegmentAllowedOperationsItem2["Read"] = "READ";
  DetailedFilterSegmentAllowedOperationsItem2["Write"] = "WRITE";
  DetailedFilterSegmentAllowedOperationsItem2["Delete"] = "DELETE";
  DetailedFilterSegmentAllowedOperationsItem2["Share"] = "SHARE";
  return DetailedFilterSegmentAllowedOperationsItem2;
})(DetailedFilterSegmentAllowedOperationsItem || {});

// packages/client/filter-segment-management/src/lib/models/filter-segment-allowed-operations-item.ts
var FilterSegmentAllowedOperationsItem = /* @__PURE__ */ ((FilterSegmentAllowedOperationsItem2) => {
  FilterSegmentAllowedOperationsItem2["Read"] = "READ";
  FilterSegmentAllowedOperationsItem2["Write"] = "WRITE";
  FilterSegmentAllowedOperationsItem2["Delete"] = "DELETE";
  FilterSegmentAllowedOperationsItem2["Share"] = "SHARE";
  return FilterSegmentAllowedOperationsItem2;
})(FilterSegmentAllowedOperationsItem || {});

// packages/client/filter-segment-management/src/lib/models/get-filter-segment-query-add-fields-item.ts
var GetFilterSegmentQueryAddFieldsItem = /* @__PURE__ */ ((GetFilterSegmentQueryAddFieldsItem2) => {
  GetFilterSegmentQueryAddFieldsItem2["Includes"] = "INCLUDES";
  GetFilterSegmentQueryAddFieldsItem2["Variables"] = "VARIABLES";
  GetFilterSegmentQueryAddFieldsItem2["Externalid"] = "EXTERNALID";
  GetFilterSegmentQueryAddFieldsItem2["Resourcecontext"] = "RESOURCECONTEXT";
  return GetFilterSegmentQueryAddFieldsItem2;
})(GetFilterSegmentQueryAddFieldsItem || {});

// packages/client/filter-segment-management/src/lib/models/get-filter-segments-query-add-fields-item.ts
var GetFilterSegmentsQueryAddFieldsItem = /* @__PURE__ */ ((GetFilterSegmentsQueryAddFieldsItem2) => {
  GetFilterSegmentsQueryAddFieldsItem2["Variables"] = "VARIABLES";
  GetFilterSegmentsQueryAddFieldsItem2["Externalid"] = "EXTERNALID";
  GetFilterSegmentsQueryAddFieldsItem2["Resourcecontext"] = "RESOURCECONTEXT";
  return GetFilterSegmentsQueryAddFieldsItem2;
})(GetFilterSegmentsQueryAddFieldsItem || {});

// packages/client/filter-segment-management/src/lib/models/get-lean-filter-segments-query-add-fields-item.ts
var GetLeanFilterSegmentsQueryAddFieldsItem = /* @__PURE__ */ ((GetLeanFilterSegmentsQueryAddFieldsItem2) => {
  GetLeanFilterSegmentsQueryAddFieldsItem2["Variables"] = "VARIABLES";
  GetLeanFilterSegmentsQueryAddFieldsItem2["Externalid"] = "EXTERNALID";
  GetLeanFilterSegmentsQueryAddFieldsItem2["Resourcecontext"] = "RESOURCECONTEXT";
  GetLeanFilterSegmentsQueryAddFieldsItem2["Owner"] = "OWNER";
  return GetLeanFilterSegmentsQueryAddFieldsItem2;
})(GetLeanFilterSegmentsQueryAddFieldsItem || {});

// packages/client/filter-segment-management/src/lib/models/lean-filter-segment-allowed-operations-item.ts
var LeanFilterSegmentAllowedOperationsItem = /* @__PURE__ */ ((LeanFilterSegmentAllowedOperationsItem2) => {
  LeanFilterSegmentAllowedOperationsItem2["Read"] = "READ";
  LeanFilterSegmentAllowedOperationsItem2["Write"] = "WRITE";
  LeanFilterSegmentAllowedOperationsItem2["Delete"] = "DELETE";
  LeanFilterSegmentAllowedOperationsItem2["Share"] = "SHARE";
  return LeanFilterSegmentAllowedOperationsItem2;
})(LeanFilterSegmentAllowedOperationsItem || {});
