// @dynatrace/strato-components/esm/filters/filter-field/decorations/decoration-marks.js
import { Decoration } from "@codemirror/view";
import clsx from "clsx";
import * as styles from "../FilterField.sty.js";
const statementMark = Decoration.mark({
  inclusive: true,
  class: styles.filterStatement,
  tagName: "span"
});
const incompleteStatementMark = Decoration.mark({
  inclusive: true,
  class: styles.incompleteFilterStatement,
  tagName: "span"
});
const intermediateStatementMark = Decoration.mark({
  inclusive: true,
  class: styles.intermediateFilterStatement,
  tagName: "span"
});
const pastingMark = Decoration.mark({
  inclusive: true,
  class: styles.pastedContent,
  tagName: "span"
});
const existsMark = Decoration.mark({
  inclusive: true,
  class: styles.existsOperator,
  tagName: "span"
});
const searchMark = Decoration.mark({
  inclusive: true,
  class: styles.operator,
  tagName: "span"
});
const keyMark = Decoration.mark({
  inclusive: true,
  class: styles.filterKey,
  tagName: "span"
});
const valueMark = Decoration.mark({
  inclusive: true,
  class: styles.filterValue,
  tagName: "span"
});
const variableMark = Decoration.mark({
  inclusive: true,
  class: styles.variable,
  tagName: "span"
});
const escapedCharacterMark = Decoration.mark({
  inclusive: true,
  class: styles.escapedCharacter,
  tagName: "span"
});
const errorMark = Decoration.mark({
  inclusive: true,
  class: styles.error,
  tagName: "span"
});
const preSpotErrorMark = Decoration.mark({
  inclusive: true,
  class: clsx(styles.spotError, styles.preSpotError),
  tagName: "span"
});
const postSpotErrorMark = Decoration.mark({
  inclusive: true,
  class: clsx(styles.spotError, styles.postSpotError),
  tagName: "span"
});
export {
  errorMark,
  escapedCharacterMark,
  existsMark,
  incompleteStatementMark,
  intermediateStatementMark,
  keyMark,
  pastingMark,
  postSpotErrorMark,
  preSpotErrorMark,
  searchMark,
  statementMark,
  valueMark,
  variableMark
};
