// @dynatrace/strato-components/esm/filters/filter-field/decorations/filter-statements.js
import { syntaxTree } from "@codemirror/language";
import { RangeSetBuilder } from "@codemirror/state";
import { ViewPlugin } from "@codemirror/view";
import {
  incompleteStatementMark,
  intermediateStatementMark,
  pastingMark,
  statementMark
} from "./decoration-marks.js";
import { pastingMode } from "../editor-state-configs/editor-config.js";
import { NEEDS_ESCAPING_REGEX } from "../utils/escape-suggestion.js";
import { getStatementStatus } from "../utils/get-statement-status.js";
const needsEscapingRegex = new RegExp(NEEDS_ESCAPING_REGEX);
function getStatementMark(statementType) {
  if (statementType === "complete") {
    return statementMark;
  }
  if (statementType === "indeterminate") {
    return intermediateStatementMark;
  }
  return incompleteStatementMark;
}
function getMergedPasteRange(statementFrom, statementTo, pastedRange, currentMerged) {
  const rangeFrom = Math.min(statementFrom, pastedRange.from);
  const rangeTo = Math.max(statementTo, pastedRange.to);
  if (currentMerged === null) {
    return { from: rangeFrom, to: rangeTo };
  }
  return {
    from: Math.min(currentMerged.from, rangeFrom),
    to: Math.max(currentMerged.to, rangeTo)
  };
}
function isPasteAdjacentToStatement(pastedRange, statementFrom, statementTo, doc, statementType) {
  if (pastedRange.from <= statementTo && pastedRange.to >= statementFrom) {
    return true;
  }
  if (statementType === "complete" && pastedRange.from > statementTo) {
    return false;
  }
  const gap = pastedRange.from > statementTo ? doc.slice(statementTo, pastedRange.from) : doc.slice(pastedRange.to, statementFrom);
  return gap.trim() === "";
}
function collectRawDecorations(view, pastedRange) {
  const decorations = [];
  let mergedPasteRange = null;
  const doc = view.state.doc.toString();
  for (const { from, to } of view.visibleRanges) {
    syntaxTree(view.state).iterate({
      from,
      to,
      enter: (node) => {
        if (node.name === "FilterStatement") {
          const { from: statementFrom, to: statementTo } = node;
          const statementType = getStatementStatus(node.node.cursor(), doc);
          const shouldMerge = pastedRange && (isPasteAdjacentToStatement(
            pastedRange,
            statementFrom,
            statementTo,
            doc,
            statementType
          ) || mergedPasteRange && isPasteAdjacentToStatement(
            mergedPasteRange,
            statementFrom,
            statementTo,
            doc,
            statementType
          ));
          if (shouldMerge) {
            mergedPasteRange = getMergedPasteRange(
              statementFrom,
              statementTo,
              pastedRange,
              mergedPasteRange
            );
          } else {
            decorations.push({
              from: statementFrom,
              to: statementTo,
              mark: getStatementMark(statementType),
              statementType
            });
          }
          return false;
        }
      }
    });
  }
  return { decorations, mergedPasteRange, doc };
}
function expandMergedRange(decorations, mergedPasteRange, doc) {
  const completeBoundaries = decorations.filter((dec) => dec.statementType === "complete").sort((a, b) => a.from - b.from);
  let leftBoundary = completeBoundaries.findLast((c) => c.to <= mergedPasteRange.from)?.to ?? 0;
  let rightBoundary = completeBoundaries.find((c) => c.from >= mergedPasteRange.to)?.from ?? doc.length;
  const leftGap = doc.slice(leftBoundary, mergedPasteRange.from);
  leftBoundary += leftGap.length - leftGap.trimStart().length;
  const rightGap = doc.slice(mergedPasteRange.to, rightBoundary);
  rightBoundary -= rightGap.length - rightGap.trimEnd().length;
  const expanded = {
    from: Math.min(mergedPasteRange.from, leftBoundary),
    to: Math.max(mergedPasteRange.to, rightBoundary)
  };
  const filtered = decorations.filter(
    (dec) => dec.statementType === "complete" || dec.to <= expanded.from || dec.from >= expanded.to
  );
  filtered.push({
    from: expanded.from,
    to: expanded.to,
    mark: incompleteStatementMark,
    statementType: "incomplete"
  });
  return { decorations: filtered, mergedPasteRange: expanded };
}
function collectStatementDecorations(view, pastedRange) {
  const rawDecorations = collectRawDecorations(view, pastedRange);
  let { decorations, mergedPasteRange } = rawDecorations;
  const { doc } = rawDecorations;
  if (!mergedPasteRange && pastedRange) {
    mergedPasteRange = { from: pastedRange.from, to: pastedRange.to };
  }
  if (mergedPasteRange) {
    const expanded = expandMergedRange(decorations, mergedPasteRange, doc);
    decorations = expanded.decorations;
    mergedPasteRange = expanded.mergedPasteRange;
  }
  return { decorations, mergedPasteRange };
}
function createFilterStatementDecorations(view) {
  const builder = new RangeSetBuilder();
  const pastingStateField = view?.state.field(pastingMode, false);
  const pastedContentHasSpecialChars = pastingStateField?.range !== void 0 && needsEscapingRegex.test(
    view.state.doc.sliceString(
      pastingStateField.range.from,
      pastingStateField.range.to
    )
  );
  const { decorations, mergedPasteRange } = collectStatementDecorations(
    view,
    pastedContentHasSpecialChars ? pastingStateField?.range : void 0
  );
  const allDecorations = decorations;
  if (pastedContentHasSpecialChars && pastingStateField?.pastingInProgress && pastingStateField.range) {
    const pastingMarkRange = mergedPasteRange ?? pastingStateField.range;
    for (const { from, to } of view.visibleRanges) {
      if (from <= pastingMarkRange.from && to >= pastingMarkRange.to) {
        allDecorations.push({
          from: pastingMarkRange.from,
          to: pastingMarkRange.to,
          mark: pastingMark,
          statementType: "complete"
        });
      }
    }
  }
  allDecorations.sort((a, b) => a.from - b.from);
  for (const decoration of allDecorations) {
    builder.add(decoration.from, decoration.to, decoration.mark);
  }
  return builder.finish();
}
const FilterStatementDecorationPlugin = ViewPlugin.fromClass(
  class {
    decorations;
    constructor(view) {
      this.decorations = createFilterStatementDecorations(view);
    }
    update(update) {
      if (update.docChanged) {
        this.decorations = createFilterStatementDecorations(update.view);
      }
    }
  },
  {
    decorations: (view) => view.decorations
  }
);
export {
  FilterStatementDecorationPlugin,
  createFilterStatementDecorations
};
