// @dynatrace/strato-components/esm/filters/filter-field/decorations/filter-statements-delete-widget.js
import { syntaxTree } from "@codemirror/language";
import { ViewPlugin, Decoration } from "@codemirror/view";
import {
  stringAnnotation,
  ANNOTATION_DELETE_STATEMENT,
  pastingMode
} from "../editor-state-configs/editor-config.js";
import { filterStatementDeleteWidget } from "../FilterField.sty.js";
import { useErrorLogsStore } from "../hooks/useErrorLogsStore.js";
import { cleanUpInput } from "../utils/delete-widget-cleanup.js";
import { NEEDS_ESCAPING_REGEX } from "../utils/escape-suggestion.js";
import { getStatementStatus } from "../utils/get-statement-status.js";
import { DeleteStatementWidget } from "../widgets/DeleteStatementWidget.js";
const needsEscapingRegex = new RegExp(NEEDS_ESCAPING_REGEX);
function createFilterStatementDeleteWidgets(view) {
  const widgets = [];
  const pastingStateField = view?.state.field(pastingMode, false);
  const pastedContentHasSpecialChars = pastingStateField?.pastingInProgress && pastingStateField.range !== void 0 && needsEscapingRegex.test(
    view.state.doc.sliceString(
      pastingStateField.range.from,
      pastingStateField.range.to
    )
  );
  const pastedRange = pastedContentHasSpecialChars ? pastingStateField.range : void 0;
  for (const { from, to } of view.visibleRanges) {
    syntaxTree(view.state).iterate({
      from,
      to,
      enter: (node) => {
        if (node.name === "FilterStatement") {
          const { from: from2, to: to2 } = node;
          if (pastedRange && pastedRange.from <= to2 && pastedRange.to >= from2) {
            return false;
          }
          const statementType = getStatementStatus(
            node.node.cursor(),
            view.state.doc.toString()
          );
          const deleteStatementWidget = Decoration.widget({
            widget: new DeleteStatementWidget(
              from2,
              to2,
              statementType === "incomplete"
            )
          });
          widgets.push(deleteStatementWidget.range(to2));
          return false;
        }
      }
    });
  }
  return Decoration.set(widgets);
}
const FilterStatementDeleteWidgetPlugin = ViewPlugin.fromClass(
  class {
    decorations;
    constructor(view) {
      this.decorations = createFilterStatementDeleteWidgets(view);
    }
    update(update) {
      if (update.docChanged) {
        this.decorations = createFilterStatementDeleteWidgets(update.view);
      }
    }
  },
  {
    decorations: (view) => view.decorations,
    eventHandlers: {
      mousedown: (event, view) => {
        const target = event.target;
        if (target.nodeName === "BUTTON" && target.parentElement && target.parentElement.classList.contains(filterStatementDeleteWidget)) {
          event.preventDefault();
          const setAction = useErrorLogsStore.getState().setAction;
          const { from: dataFrom, to: dataTo } = target.dataset;
          const parsedFrom = parseInt(dataFrom ?? "0");
          const parsedTo = parseInt(dataTo ?? "0");
          const currentValue = view.state.doc.toString();
          const adjustedRange = cleanUpInput(view.state, {
            from: parsedFrom,
            to: parsedTo
          });
          const { from, to } = adjustedRange;
          setAction({
            type: "deleteStatement",
            data: {
              value: currentValue,
              range: { from, to },
              newCursorPos: from
            }
          });
          view.dispatch({
            changes: {
              from,
              to,
              insert: ""
            },
            annotations: stringAnnotation.of(ANNOTATION_DELETE_STATEMENT),
            selection: { anchor: from }
          });
        }
      }
    }
  }
);
export {
  FilterStatementDeleteWidgetPlugin,
  createFilterStatementDeleteWidgets
};
