// @dynatrace/strato-components/charts/histogram/HistogramChartInternal.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var HistogramChartInternal_exports = {};
__export(HistogramChartInternal_exports, {
  HistogramChartInternal: () => HistogramChartInternal
});
module.exports = __toCommonJS(HistogramChartInternal_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_constants = require("./components/annotations-histogram/constants.js");
var import_HistogramAnnotationsArea = require("./components/annotations-histogram/HistogramAnnotationsArea.js");
var import_HistogramAnnotationsIndicators = require("./components/annotations-histogram/HistogramAnnotationsIndicators.js");
var import_HistogramValueBottomAxisArea = require("./components/axes/HistogramValueBottomAxisArea/HistogramValueBottomAxisArea.js");
var import_HistogramValueLeftAxisArea = require("./components/axes/HistogramValueLeftAxisArea.js");
var import_HistogramValueRightAxisArea = require("./components/axes/HistogramValueRightAxisArea.js");
var import_HistogramChartPlot = require("./components/layout/HistogramChartPlot.js");
var import_HistogramOverlays = require("./components/overlays/HistogramOverlays.js");
var import_HistogramThresholdArea = require("./components/thresholds/HistogramThresholdArea.js");
var import_ThresholdIndicators = require("./components/thresholds/ThresholdIndicators.js");
var import_defaults = require("./constants/defaults.js");
var import_histogram_grid_area_rules = require("./constants/histogram-grid-area-rules.js");
var import_useHistogramAnnotationsConfig = require("./hooks/useHistogramAnnotationsConfig.js");
var import_useHistogramFilteredData = require("./hooks/useHistogramFilteredData.js");
var import_useHistogramLegend = require("./hooks/useHistogramLegend.js");
var import_useHistogramThresholds = require("./hooks/useHistogramThresholds.js");
var import_useYAxisConfig = require("./hooks/useYAxisConfig.js");
var import_HistogramAnnotationsDataProvider = require("./providers/HistogramAnnotationsDataProvider.js");
var import_HistogramChartImperativeHandler = require("./providers/HistogramChartImperativeHandler.js");
var import_HistogramFormatterProvider = require("./providers/HistogramFormatterProvider.js");
var import_HistogramXScaleProvider = require("./providers/HistogramXScaleProvider.js");
var import_HistogramYScalesProvider = require("./providers/HistogramYScalesProvider.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_defaultChartValues_css = require("../../core/styles/defaultChartValues.sty.js");
var import_constants2 = require("../core/components/annotations/constants.js");
var import_get_annotations_visible_area = require("../core/components/annotations/utils/get-annotations-visible-area.js");
var import_ChartGridRoot = require("../core/components/chart-grid/ChartGridRoot.js");
var import_useChartLayoutLoading = require("../core/components/chart-layout/hooks/useChartLayoutLoading.js");
var import_useChartLayoutOrientation = require("../core/components/chart-layout/hooks/useChartLayoutOrientation.js");
var import_useGraphSize = require("../core/components/chart-layout/hooks/useGraphSize.js");
var import_ChartLoaderOverlay = require("../core/components/chart-loader-overlay/ChartLoaderOverlay.js");
var import_constants3 = require("../core/components/data-point/constants.js");
const histogramGridTemplateAreas = `
  "${import_histogram_grid_area_rules.AREAS.annotations} ${import_histogram_grid_area_rules.AREAS.annotations} ${import_histogram_grid_area_rules.AREAS.annotations} . ."
  "${import_histogram_grid_area_rules.AREAS.leftAxis} ${import_histogram_grid_area_rules.AREAS.leftThreshold} plot ${import_histogram_grid_area_rules.AREAS.rightThreshold} ${import_histogram_grid_area_rules.AREAS.rightAxis}"
  ". . ${import_histogram_grid_area_rules.AREAS.bottomAxis} . ."
`;
const histogramGridTemplateColumns = "minmax(0, auto) auto minmax(60%, 1fr) auto minmax(0, auto)";
const HistogramChartInternal = (0, import_react.forwardRef)(({ config }, forwardedRef) => {
  const loading = (0, import_useChartLayoutLoading.useChartLayoutLoading)();
  const annotations = (0, import_useHistogramAnnotationsConfig.useHistogramAnnotationsConfig)();
  const tracksQuantity = annotations.tracks.length;
  const numHiddenTracks = annotations.tracks.filter(
    (value) => value.hidden
  ).length;
  const annotationsVisibleArea = (0, import_get_annotations_visible_area.getAnnotationsVisibleArea)(
    tracksQuantity,
    annotations?.visibleTracksLimit,
    numHiddenTracks
  );
  const histogramGridTemplateRows = `minmax(0, ${annotationsVisibleArea + (tracksQuantity ? import_constants.ANNOTATIONS_PADDING_BOTTOM : 0)}px) minmax(60%, 1fr) minmax(0, auto)`;
  const { leftThreshold, rightThreshold } = (0, import_useHistogramThresholds.useHistogramThresholds)();
  const hasThresholds = !(0, import_lodash_es.isEmpty)(leftThreshold) || !(0, import_lodash_es.isEmpty)(rightThreshold);
  const { width, height } = (0, import_useGraphSize.useGraphSize)();
  const chartRef = (0, import_react.useRef)(null);
  const mergedRef = (0, import_useMergeRefs.useMergeRefs)([
    forwardedRef,
    chartRef
  ]);
  const { orientation } = (0, import_useChartLayoutOrientation.useChartLayoutOrientation)();
  const { right: axisRight } = (0, import_useYAxisConfig.useYAxisConfig)();
  const { hidden: legendHidden } = (0, import_useHistogramLegend.useHistogramLegend)();
  const { units } = (0, import_useHistogramFilteredData.useHistogramFilteredData)();
  const availableWidth = tracksQuantity && orientation === "vertical" ? width - import_constants2.SCROLLBAR_WIDTH : width;
  const noAxisRight = !axisRight || units.length === 1;
  const noLegendRight = legendHidden || orientation === "vertical";
  const noLegendBottom = legendHidden || orientation === "horizontal";
  const hasRightPadding = noAxisRight && noLegendRight;
  const chartGridHeight = height - (!tracksQuantity || noLegendBottom ? import_constants3.OVERLAY_PADDING + import_constants3.BOTTOM_PADDING_ADJUSTMENT : 0);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      style: {
        paddingTop: !tracksQuantity ? import_constants3.OVERLAY_PADDING : 0,
        paddingLeft: 0
      },
      className: import_defaultChartValues_css.defaultChartSize,
      role: "img",
      "aria-label": import_defaults.HISTOGRAM_CHART_ARIA_LABEL,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_ChartGridRoot.ChartGrid,
        {
          width: availableWidth - (hasRightPadding ? import_constants3.OVERLAY_PADDING : 0),
          height: chartGridHeight,
          gridTemplateAreas: histogramGridTemplateAreas,
          gridTemplateColumns: histogramGridTemplateColumns,
          gridTemplateRows: histogramGridTemplateRows,
          ref: mergedRef,
          "data-testid": "histogram-chart-root",
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_HistogramXScaleProvider.HistogramXScaleProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_HistogramYScalesProvider.HistogramYScalesProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_HistogramFormatterProvider.HistogramFormatterProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_HistogramAnnotationsDataProvider.HistogramAnnotationsDataProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
            import_HistogramChartImperativeHandler.HistogramChartImperativeHandler,
            {
              config,
              forwardedRef,
              chartRef,
              children: [
                /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_HistogramValueLeftAxisArea.HistogramValueLeftAxisArea, {}),
                /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_HistogramValueBottomAxisArea.HistogramValueBottomAxisArea, {}),
                /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_HistogramValueRightAxisArea.HistogramValueRightAxisArea, {}),
                !!tracksQuantity && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_HistogramAnnotationsArea.HistogramAnnotationsArea, {}),
                /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_ChartGridRoot.ChartGrid.PlotArea, { children: [
                  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_HistogramChartPlot.HistogramChartPlot, {}),
                  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_HistogramAnnotationsIndicators.HistogramAnnotationsIndicators, {}),
                  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ThresholdIndicators.ThresholdIndicators, {}),
                  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                    import_ChartLoaderOverlay.ChartLoaderOverlay,
                    {
                      loading: !!loading,
                      zoomEnabled: !!config.interactions?.zoom?.enabled,
                      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_HistogramOverlays.HistogramOverlays, {})
                    }
                  )
                ] }),
                hasThresholds && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_HistogramThresholdArea.HistogramThresholdArea, {})
              ]
            }
          ) }) }) }) })
        }
      )
    }
  );
});
HistogramChartInternal.displayName = import_defaults.DISPLAY_NAME;
