// @dynatrace/strato-components/charts/histogram/HistogramChartRoot.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var HistogramChartRoot_exports = {};
__export(HistogramChartRoot_exports, {
  HistogramChartRoot: () => HistogramChartRoot
});
module.exports = __toCommonJS(HistogramChartRoot_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_HistogramChartLegend = require("./components/legend/HistogramChartLegend.js");
var import_defaults = require("./constants/defaults.js");
var import_HistogramChartInternal = require("./HistogramChartInternal.js");
var import_useHistogramChartConfig = require("./hooks/useHistogramChartConfig.js");
var import_HistogramConfigProvider = require("./providers/HistogramConfigProvider.js");
var import_HistogramDataProvider = require("./providers/HistogramDataProvider.js");
var import_HistogramFilteredDataProvider = require("./providers/HistogramFilteredDataProvider.js");
var import_HistogramStateSyncProvider = require("./providers/HistogramStateSyncProvider.js");
var import_StoreProvider = require("./providers/StoreProvider.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_ChartLayout = require("../core/components/chart-layout/ChartLayout.js");
var import_useAutoLegendRefresh = require("../core/components/chart-layout/hooks/useAutoLegendRefresh.js");
var import_get_chart_states_template = require("../core/components/chart-states/utils/get-chart-states-template.js");
var import_OverlayContext = require("../core/components/overlay-tooltip/contexts/OverlayContext.js");
var import_TooltipContext = require("../core/components/overlay-tooltip/contexts/TooltipContext.js");
var import_ChartSeriesActionContext = require("../core/components/series-action/ChartSeriesActionContext.js");
var import_coerceSizeValue = require("../core/utils/coerceSizeValue.js");
const COMPONENT_NAME_BEHAVIOURAL_TRACKING = "HistogramChart";
const HistogramChartRoot = (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    children,
    data,
    height,
    width,
    colorPalette,
    truncationMode,
    loading = import_defaults.DEFAULT_LOADING_STATE,
    seriesActions,
    initialZoom,
    currentZoom,
    onZoomChange,
    infiniteZoom,
    className,
    style,
    valueRepresentation,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow
  } = props;
  const [behaviorTrackingProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(props);
  const chartHeight = (0, import_coerceSizeValue.coerceSizeValue)(height ?? import_defaults.DEFAULT_CHART_HEIGHT);
  const chartConfig = (0, import_useHistogramChartConfig.useHistogramChartConfig)(children, {
    colorPalette,
    truncationMode,
    valueRepresentation,
    initialZoom,
    currentZoom,
    onZoomChange,
    infiniteZoom
  });
  const { emptyState, errorState } = (0, import_get_chart_states_template.getChartStatesTemplates)(props.children);
  const {
    position: legendPosition,
    onRatioChange,
    // eslint-disable-next-line @typescript-eslint/no-deprecated
    ratio,
    initialRatio,
    hidden: isLegendHidden
  } = chartConfig.legend;
  const legendRatio = initialRatio ?? ratio;
  const chartLayoutRef = (0, import_useAutoLegendRefresh.useAutoLegendRefresh)(data);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_HistogramConfigProvider.HistogramConfigProvider, { config: chartConfig, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_HistogramDataProvider.HistogramDataProvider, { data, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ChartSeriesActionContext.ChartSeriesActionContext.Provider,
    {
      value: {
        // We need a cast here as the provider expects unknown and the
        // react context will not be able to carry the generic anyways.
        chartSeriesActions: seriesActions
      },
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_StoreProvider.StoreProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_HistogramFilteredDataProvider.HistogramFilteredDataProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_HistogramStateSyncProvider.HistogramStateSyncProvider,
        {
          currentDomain: chartConfig.currentZoom,
          initialDomain: chartConfig.initialZoom,
          infiniteZoom: chartConfig.infiniteZoom,
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayContext.OverlayProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TooltipContext.TooltipProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
            import_ChartLayout.ChartLayout,
            {
              ref: chartLayoutRef,
              chartHeight,
              chartWidth: width,
              isEmpty: (0, import_lodash_es.isEmpty)(data),
              emptyState,
              errorState,
              showLoader: loading,
              className,
              style,
              ...behaviorTrackingProps,
              "data-dtrum-mask": dataDtrumMask,
              "data-dtrum-allow": dataDtrumAllow,
              "data-dt-component": COMPONENT_NAME_BEHAVIOURAL_TRACKING,
              children: [
                /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ChartLayout.ChartLayout.Graph, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  import_HistogramChartInternal.HistogramChartInternal,
                  {
                    ref: forwardedRef,
                    config: chartConfig
                  }
                ) }),
                !isLegendHidden && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  import_ChartLayout.ChartLayout.Legend,
                  {
                    position: legendPosition,
                    initialRatio: legendRatio,
                    onResize: onRatioChange,
                    children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_HistogramChartLegend.HistogramChartLegend, {})
                  }
                )
              ]
            }
          ) }) })
        }
      ) }) })
    }
  ) }) });
});
HistogramChartRoot.displayName = "HistogramChartRoot";
