// @dynatrace/strato-components/tables/DataTable/components/Toolbar/Download/DataTableDownloadData.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DataTableDownloadData_exports = {};
__export(DataTableDownloadData_exports, {
  DataTableDownloadData: () => DataTableDownloadData
});
module.exports = __toCommonJS(DataTableDownloadData_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_intl = require("react-intl");
var styles = __toESM(require("./DataTableDownloadData.sty.js"));
var import_Menu = require("../../../../../navigation/menu/Menu.js");
var import_DataTableContext = require("../../../contexts/DataTableContext.js");
var import_messages = require("../../../messages.js");
const DataTableDownloadData = ({
  excludeColumns,
  hasIntents,
  isDownloading,
  onDownloadStart,
  onDownloadEnd
}) => {
  const intl = (0, import_react_intl.useIntl)();
  const table = (0, import_DataTableContext.useDataTable)();
  const selectionCount = table.getSelectionCount();
  const menuGroupLabel = intl.formatMessage(import_messages.dataTableMessages.downloadTable);
  const downloadMenuItems = /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      import_Menu.Menu.Item,
      {
        onSelect: () => {
          if (isDownloading) {
            return;
          }
          onDownloadStart();
          table.downloadAllData(excludeColumns, onDownloadEnd);
        },
        className: styles.downloadMenuItem,
        "data-dt-name": "ToolbarDownloadAll",
        children: [
          intl.formatMessage(import_messages.dataTableMessages.downloadAll),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Suffix, { className: styles.downloadMenuSuffix, children: intl.formatMessage(import_messages.dataTableMessages.downloadRowsCount, {
            noOfRows: table.getPrePaginationRowModel().flatRows.length
          }) })
        ]
      }
    ),
    table.options.enablePagination && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      import_Menu.Menu.Item,
      {
        onSelect: () => {
          if (isDownloading) {
            return;
          }
          onDownloadStart();
          table.downloadPageData(excludeColumns, onDownloadEnd);
        },
        className: styles.downloadMenuItem,
        "data-dt-name": "ToolbarDownloadPage",
        children: [
          intl.formatMessage(import_messages.dataTableMessages.downloadPage),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Suffix, { className: styles.downloadMenuSuffix, children: intl.formatMessage(import_messages.dataTableMessages.downloadRowsCount, {
            noOfRows: table.getPaginationRowModel().flatRows.length
          }) })
        ]
      }
    ),
    table.options.enableRowSelection && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      import_Menu.Menu.Item,
      {
        disabled: selectionCount === 0,
        onSelect: () => {
          if (isDownloading) {
            return;
          }
          onDownloadStart();
          table.downloadSelectedData(excludeColumns, onDownloadEnd);
        },
        className: styles.downloadMenuItem,
        "data-dt-name": "ToolbarDownloadSelected",
        children: [
          intl.formatMessage(import_messages.dataTableMessages.downloadSelected),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Suffix, { className: styles.downloadMenuSuffix, children: intl.formatMessage(import_messages.dataTableMessages.downloadRowsCount, {
            noOfRows: selectionCount
          }) })
        ]
      }
    )
  ] });
  return hasIntents ? /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Menu.Menu.Sub, { defaultOpen: false, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.SubTrigger, { children: menuGroupLabel }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.SubContent, { children: downloadMenuItems })
  ] }) : /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Label, { children: menuGroupLabel }),
    downloadMenuItems
  ] });
};
