// @dynatrace/strato-components/esm/charts/core/components/chart-layout/ChartLayout.js
import { jsx } from "react/jsx-runtime";
import {
  forwardRef,
  useImperativeHandle,
  useRef
} from "react";
import { LayoutStateHandler } from "./components/LayoutStateHandler.js";
import { ResizableChartLayout } from "./components/ResizableChartLayout.js";
import { ChartLayoutOrientationContext } from "./context/ChartLayoutOrientation.context.js";
import { ChartLayoutRatioContext } from "./context/ChartLayoutRatio.context.js";
import { ChartLayoutSizeContext } from "./context/ChartLayoutSize.context.js";
import { ChartLayoutTemplateContext } from "./context/ChartLayoutTemplate.context.js";
import { useAutomaticRatio } from "./hooks/useAutomaticRatio.js";
import { Graph } from "./slots/Graph.js";
import { Legend } from "./slots/Legend.js";
import { Provider } from "./slots/Provider.js";
import { getChartLayoutSettings } from "./utils/get-chart-layout-settings.js";
import { getLayoutGridSettings } from "./utils/get-layout-grid-settings.js";
import { useBehavioralTrackingProps } from "../../../../core/hooks/useBehavioralTrackingProps.js";
import { useMergeRefs } from "../../../../core/hooks/useMergeRefs.js";
import { Grid } from "../../../../layouts/grid/Grid.js";
import { DataDtrumContext } from "../../context/DataDtrumContext.js";
import { useObserveContainerSize } from "../../hooks/useObserveContainerSize.js";
const ChartLayoutRoot = forwardRef((props, forwardedRef) => {
  const {
    chartHeight,
    chartWidth,
    children,
    isEmpty = false,
    emptyState,
    errorState,
    showLoader = false,
    className,
    style,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "data-testid": dataTestId = "chart-layout",
    "data-dt-component": dataDtComponent,
    "aria-label": ariaLabel,
    role
  } = props;
  const [behavioralTrackingProps] = useBehavioralTrackingProps(props);
  const [containerSizes, gridContainer] = useObserveContainerSize();
  const containerAvailableArea = {
    width: containerSizes.width,
    height: containerSizes.height,
    wrapperHeight: containerSizes.wrapperHeight
  };
  const chartLayoutRef = useRef(null);
  const mergedRefs = useMergeRefs([
    gridContainer,
    chartLayoutRef
  ]);
  const {
    graphArea,
    legendArea,
    legendSettings,
    provider: Provider2
  } = getChartLayoutSettings(containerSizes, children, isEmpty);
  const {
    hidden,
    orientation,
    ratio,
    initialRatio,
    overrideRatio,
    onResize,
    showResizer,
    ratioBoundaries
  } = legendSettings;
  const internalRatioRef = useRef(void 0);
  if (internalRatioRef.current === void 0 && legendArea !== void 0) {
    internalRatioRef.current = initialRatio ?? ratio;
  }
  const legendRatio = internalRatioRef.current ?? ratio;
  const {
    legendCandidateSize,
    readyForAutoRatio,
    legendRef,
    refreshAutoRatio
  } = useAutomaticRatio(orientation, legendRatio, containerAvailableArea);
  const { layoutWidth, layoutHeight, templateAreas, gridTemplate } = getLayoutGridSettings(
    chartWidth,
    chartHeight,
    containerSizes,
    !hidden,
    showResizer
  );
  useImperativeHandle(
    forwardedRef,
    () => ({
      element: chartLayoutRef?.current,
      refreshLegendSize: refreshAutoRatio
    })
  );
  return /* @__PURE__ */ jsx(ChartLayoutTemplateContext.Provider, { value: { graphArea, legendArea }, children: /* @__PURE__ */ jsx(ChartLayoutOrientationContext.Provider, { value: { orientation }, children: /* @__PURE__ */ jsx(
    ChartLayoutRatioContext.Provider,
    {
      value: {
        legendRef,
        ratio: legendRatio,
        ratioBoundaries,
        overrideRatio,
        readyForAutoRatio,
        legendCandidateSize
      },
      children: /* @__PURE__ */ jsx(ChartLayoutSizeContext.Provider, { value: containerAvailableArea, children: /* @__PURE__ */ jsx(
        DataDtrumContext.Provider,
        {
          value: { dataDtrumAllow, dataDtrumMask },
          children: /* @__PURE__ */ jsx(
            Grid,
            {
              ...behavioralTrackingProps,
              "data-testid": dataTestId,
              "data-dtrum-mask": dataDtrumMask,
              "data-dtrum-allow": dataDtrumAllow,
              "data-dt-component": dataDtComponent,
              className,
              width: layoutWidth,
              height: layoutHeight,
              ref: mergedRefs,
              gridTemplateColumns: gridTemplate,
              gridTemplateRows: gridTemplate,
              gridTemplateAreas: templateAreas,
              gap: 0,
              "aria-label": ariaLabel,
              role,
              style: {
                position: "relative",
                overflow: "hidden",
                ...style
              },
              children: /* @__PURE__ */ jsx(
                LayoutStateHandler,
                {
                  empty: isEmpty,
                  height: layoutHeight,
                  emptyState,
                  errorState,
                  showLoader,
                  children: /* @__PURE__ */ jsx(Provider2, { children: /* @__PURE__ */ jsx(
                    ResizableChartLayout,
                    {
                      onResize,
                      showResizer,
                      showLegend: !hidden
                    }
                  ) })
                }
              )
            }
          )
        }
      ) })
    }
  ) }) });
});
ChartLayoutRoot.displayName = "ChartLayoutRoot";
const ChartLayout = Object.assign(ChartLayoutRoot, {
  Graph,
  Legend,
  Provider
});
export {
  ChartLayout,
  ChartLayoutRoot
};
