// @dynatrace/strato-components/esm/filters/filter-field/utils/suggestions-utils/get-suggestions.js
import {
  getKeySuggestions,
  JSONPATH_PATTERN_PARTIAL
} from "./key-suggestions.js";
import {
  defaultLogicalOperatorSuggestions,
  getComparisonOperatorSuggestions
} from "./operator-suggestions.js";
import { getJSONPathHint, getPastingSuggestion } from "./shared-suggestions.js";
import {
  getSuggestionMatchPriority,
  normalizeSuggestionMatchText,
  suggestionMatchesTerm
} from "./suggestion-matching.js";
import { getValidValueTypes } from "./suggestion-type.js";
import {
  getValueSuggestionGroups,
  VALUE_GROUP_ID
} from "./value-suggestions.js";
import { NEEDS_ESCAPING_REGEX } from "../escape-suggestion.js";
import { KEY_GROUP_ID } from "../validation-utils/normalize-validator-map.js";
import { isWildCardOperator } from "../validation-utils/validation-helper-functions.js";
const comparisonOperatorAlias = {
  "=*": "= *",
  "!=*": "!= *"
};
function getCurrentValueFromToken(token) {
  return token.type === "String" || token.type === "Key" ? token.value : token.textValue;
}
function getValueSuggestionsTypesFromKeyPredicate(internalValidatorMap, currentStatement) {
  if (!internalValidatorMap || !currentStatement?.key?.value) {
    return [];
  }
  const keyConfig = internalValidatorMap.normalizedKeys.byKey.get(currentStatement.key.value);
  if (!keyConfig) {
    return [];
  }
  if (keyConfig.valueType) {
    return getValidValueTypes({ type: keyConfig.valueType }).types;
  }
  return getValidValueTypes(keyConfig.valuePredicate).types;
}
function getTypedOperatorFromCurrentStatement(currentStatement) {
  const needle = currentStatement?.comparisonOperator ? currentStatement.comparisonOperator.textValue : "";
  return comparisonOperatorAlias[needle] ?? needle;
}
function getSuggestions({
  suggestionTypes,
  overlappingSuggestions,
  currentStatement,
  currentToken,
  currentTokenType,
  prevNode,
  internalValidatorMap,
  customTypes,
  parserConfig,
  showLogicalOperators,
  showInclusionOperators,
  doc,
  suggestionTypeForToken
}) {
  const groupedSuggestions = {};
  const currentValue = getCurrentValueFromToken(currentToken);
  let pastedContentSuggestions = [];
  const shouldSortAlphabetically = internalValidatorMap?.sortSuggestions ?? true;
  for (const { type: suggestionType, token } of suggestionTypeForToken ?? []) {
    if (suggestionType === "key") {
      const currentValue2 = getCurrentValueFromToken(token);
      const keySuggestionGroups = getKeySuggestions(
        internalValidatorMap,
        suggestionTypes,
        customTypes,
        doc,
        token
      );
      assignGroupsToSlots({
        groups: keySuggestionGroups,
        groupedSuggestions,
        slotName: "key",
        implicitIdPrefix: KEY_GROUP_ID,
        token,
        currentValue: currentValue2,
        shouldSortAlphabetically
      });
    } else if (suggestionType === "value") {
      const showDefaultValueOrWildcardSuggestions = typeof currentStatement?.comparisonOperator?.value === "string" && currentStatement?.comparisonOperator?.type !== "Error" && isWildCardOperator(currentStatement);
      const showWildcardValueSuggestions = prevNode === "ExistsOperator" && currentTokenType === "comparisonOperator" || isWildCardOperator(currentStatement);
      const valueSuggestionGroups = getValueSuggestionGroups(
        currentStatement?.key?.value,
        currentStatement?.comparisonOperator?.value?.toString(),
        internalValidatorMap,
        token,
        showDefaultValueOrWildcardSuggestions,
        showWildcardValueSuggestions,
        suggestionTypes.includes("pastedContent")
      );
      const valueSuggestionsTypes = getValueSuggestionsTypesFromKeyPredicate(
        internalValidatorMap,
        currentStatement
      );
      assignGroupsToSlots({
        groups: valueSuggestionGroups,
        groupedSuggestions,
        slotName: "value",
        implicitIdPrefix: VALUE_GROUP_ID,
        token,
        currentValue,
        shouldSortAlphabetically,
        suggestionsTypeOrder: valueSuggestionsTypes
      });
    } else if (suggestionType === "pastedContent" && token.textValue?.match(NEEDS_ESCAPING_REGEX)) {
      pastedContentSuggestions = [getPastingSuggestion()];
    }
  }
  if (suggestionTypes.includes("comparisonOperator") || showInclusionOperators) {
    let key = currentStatement?.key?.textValue ?? "";
    if (overlappingSuggestions) {
      const normalizedKeyEnd = overlappingSuggestions.subToken.range.from - (currentStatement?.range.from ?? 0);
      key = key.slice(0, normalizedKeyEnd).trim();
    }
    let comparisonOperatorSuggestions = getComparisonOperatorSuggestions(
      key,
      internalValidatorMap,
      parserConfig
    );
    comparisonOperatorSuggestions = filterSuggestions(
      comparisonOperatorSuggestions,
      currentToken,
      overlappingSuggestions ? overlappingSuggestions.subToken.textValue.trim() : getTypedOperatorFromCurrentStatement(currentStatement),
      // Operator details (e.g. "equals", "not equals") are display-only
      // localized labels — never a match target.
      { includeDetails: false }
    );
    if (overlappingSuggestions) {
      comparisonOperatorSuggestions = comparisonOperatorSuggestions.map(
        (suggestion) => ({
          ...suggestion,
          insertionRange: overlappingSuggestions.subToken.range,
          escapeSuggestion: false
        })
      );
    } else {
      comparisonOperatorSuggestions = comparisonOperatorSuggestions.map(
        (suggestion) => ({
          ...suggestion,
          insertionRange: void 0,
          escapeSuggestion: false
        })
      );
    }
    if (comparisonOperatorSuggestions.length > 0 && !suggestionTypes.includes("pastedContent")) {
      groupedSuggestions.comparisonOperator = {
        id: "strato-filter-field-comparator-suggestions",
        suggestions: comparisonOperatorSuggestions
      };
    }
  }
  let logicalOperatorSuggestions = [];
  if (showLogicalOperators) {
    logicalOperatorSuggestions = filterSuggestions(
      defaultLogicalOperatorSuggestions,
      currentToken,
      currentValue,
      { includeDetails: false }
    );
  }
  if (suggestionTypes.find((value) => value === "key") && (currentToken.type === "JSONPath" || currentToken.type === "Error" && currentToken.textValue.match(JSONPATH_PATTERN_PARTIAL))) {
    groupedSuggestions.key = {
      id: "strato-filter-field-key-suggestions",
      suggestions: [getJSONPathHint()]
    };
  }
  return {
    groupedSuggestions,
    logicalOperatorSuggestions,
    pastedContentSuggestions
  };
}
function trimEscapedCharacters(value) {
  return normalizeSuggestionMatchText(value);
}
function filterSuggestions(providedSuggestions, currentToken, currentValue, options) {
  const textValue = trimEscapedCharacters(currentValue);
  return providedSuggestions.filter((suggestion) => {
    if (Array.isArray(currentToken.value)) {
      return true;
    }
    if (!suggestion.value) {
      return true;
    }
    if (suggestion.value === "OR" && currentToken.value !== currentToken.textValue) {
      return false;
    }
    return suggestionMatchesTerm(suggestion, textValue, options);
  });
}
function isSuggestionTypeNumber(suggestion) {
  if (suggestion.valueType) {
    return suggestion.valueType === "Number";
  }
  const value = trimEscapedCharacters(suggestion.value ?? "");
  return value ? value.trim() !== "" && !isNaN(Number(value)) : false;
}
function isSuggestionTypeDuration(suggestion) {
  if (suggestion.valueType) {
    return suggestion.valueType === "Duration";
  }
  const value = trimEscapedCharacters(suggestion.value ?? "");
  return /^\d+(ns|ms|s|m|h|d)$/.test(value ?? "");
}
function getDurationNumericValue(value) {
  const match = /^(\d+)(ns|ms|s|m|h|d)$/.exec(value);
  return match ? Number(match[1]) : 0;
}
function getDurationUnit(value) {
  const match = /^\d+(ns|ms|s|m|h|d)$/.exec(value);
  return match ? match[1] : "";
}
const DURATION_UNIT_ORDER = {
  ns: 0,
  ms: 1,
  s: 2,
  m: 3,
  h: 4,
  d: 5
};
function compareDurations(a, b) {
  const unitA = getDurationUnit(a);
  const unitB = getDurationUnit(b);
  const unitDiff = (DURATION_UNIT_ORDER[unitA] ?? 0) - (DURATION_UNIT_ORDER[unitB] ?? 0);
  if (unitDiff !== 0) {
    return unitDiff;
  }
  return getDurationNumericValue(a) - getDurationNumericValue(b);
}
function sortSuggestions(suggestions, isSortingEnabled, currentValue, suggestionsTypeOrder) {
  const { validatorMapSuggestions, otherSuggestions } = suggestions.reduce(
    (acc, currentSuggestion) => {
      if (currentSuggestion.suggestionType === "validatorMap") {
        acc.validatorMapSuggestions.push(currentSuggestion);
      } else {
        acc.otherSuggestions.push(currentSuggestion);
      }
      return acc;
    },
    { validatorMapSuggestions: [], otherSuggestions: [] }
  );
  currentValue = currentValue ? trimEscapedCharacters(currentValue) : "";
  const shouldSortMixedSuggestionsTypes = isSortingEnabled && suggestionsTypeOrder && suggestionsTypeOrder.length > 1;
  if (shouldSortMixedSuggestionsTypes) {
    return [
      ...sortMixedSuggestionTypes(
        validatorMapSuggestions,
        suggestionsTypeOrder,
        currentValue
      ),
      ...otherSuggestions
    ];
  }
  if (currentValue || isSortingEnabled) {
    return [
      ...sortSuggestionsGroup(validatorMapSuggestions, currentValue),
      ...otherSuggestions
    ];
  }
  return [...validatorMapSuggestions, ...otherSuggestions];
}
function sortMixedSuggestionTypes(validatorMapSuggestions, suggestionsTypeOrder, currentValue) {
  const suggestionsGroups = /* @__PURE__ */ new Map();
  const ungroupedSuggestions = [];
  for (const suggestion of validatorMapSuggestions) {
    const { valueType } = suggestion;
    if (valueType !== void 0 && suggestionsTypeOrder.includes(valueType)) {
      let suggestionGroup = suggestionsGroups.get(valueType);
      if (!suggestionGroup) {
        suggestionGroup = [];
        suggestionsGroups.set(valueType, suggestionGroup);
      }
      suggestionGroup.push(suggestion);
    } else {
      ungroupedSuggestions.push(suggestion);
    }
  }
  const sortedResult = [];
  for (const suggestionType of suggestionsTypeOrder) {
    sortedResult.push(
      ...sortSuggestionsGroup(
        suggestionsGroups.get(suggestionType) ?? [],
        currentValue
      )
    );
  }
  sortedResult.push(
    ...sortSuggestionsGroup(ungroupedSuggestions, currentValue)
  );
  return sortedResult;
}
function sortSuggestionsGroup(suggestionsGroup, currentValue) {
  if (currentValue) {
    return suggestionsGroup.sort((a, b) => {
      const aNorm = trimEscapedCharacters(a.value ?? "");
      const bNorm = trimEscapedCharacters(b.value ?? "");
      const priorityDiff = getSuggestionMatchPriority(a, currentValue) - getSuggestionMatchPriority(b, currentValue);
      if (priorityDiff !== 0) {
        return priorityDiff;
      }
      if (isSuggestionTypeNumber(a) && isSuggestionTypeNumber(b)) {
        return Number(aNorm) - Number(bNorm);
      }
      if (isSuggestionTypeDuration(a) && isSuggestionTypeDuration(b)) {
        return compareDurations(aNorm, bNorm);
      }
      return (a.value ?? "").localeCompare(b.value ?? "");
    });
  }
  const { numericSuggestions, durationSuggestions, nonNumericSuggestions } = suggestionsGroup.reduce(
    (acc, currentSuggestion) => {
      if (isSuggestionTypeNumber(currentSuggestion)) {
        acc.numericSuggestions.push(currentSuggestion);
      } else if (isSuggestionTypeDuration(currentSuggestion)) {
        acc.durationSuggestions.push(currentSuggestion);
      } else {
        acc.nonNumericSuggestions.push(currentSuggestion);
      }
      return acc;
    },
    {
      numericSuggestions: [],
      durationSuggestions: [],
      nonNumericSuggestions: []
    }
  );
  numericSuggestions.sort(
    (a, b) => Number(a.value ?? 0) - Number(b.value ?? 0)
  );
  durationSuggestions.sort(
    (a, b) => compareDurations(a.value ?? "", b.value ?? "")
  );
  nonNumericSuggestions.sort(
    (a, b) => (a.value ?? "").localeCompare(b.value ?? "")
  );
  return [
    ...numericSuggestions,
    ...durationSuggestions,
    ...nonNumericSuggestions
  ];
}
function assignGroupsToSlots({
  groups,
  groupedSuggestions,
  slotName,
  implicitIdPrefix,
  token,
  currentValue,
  shouldSortAlphabetically,
  suggestionsTypeOrder
}) {
  for (const group of groups) {
    const filtered = filterSuggestions(group.suggestions, token, currentValue);
    if (filtered.length === 0) {
      continue;
    }
    const sorted = sortSuggestions(
      filtered,
      shouldSortAlphabetically,
      currentValue,
      suggestionsTypeOrder
    );
    if (!groupedSuggestions[slotName] && (group.id === implicitIdPrefix || group.id.startsWith(`${implicitIdPrefix}-implicit`))) {
      groupedSuggestions[slotName] = { id: group.id, suggestions: sorted };
    } else {
      groupedSuggestions[group.id] = {
        id: group.id,
        label: group.label,
        "aria-label": group["aria-label"],
        suggestions: sorted
      };
    }
  }
}
export {
  filterSuggestions,
  getSuggestions
};
