// @dynatrace/strato-components/esm/filters/filter-field/utils/suggestions-utils/get-current-token.js
import {
  isFilterFieldLeafNode,
  isFilterFieldListNode,
  isFilterFieldValueNode
} from "../tree-node-guards.js";
import { getFilterStatement } from "./suggestions-callback.js";
function getCurrentToken({
  tree,
  convertedTree,
  cursorPosition,
  doc
}) {
  const fallbackToken = {
    type: "String",
    textValue: "",
    value: "",
    range: {
      from: cursorPosition,
      to: cursorPosition
    }
  };
  let currentToken = fallbackToken;
  let currentTokenType = void 0;
  const currentStatement = getFilterStatement(
    convertedTree,
    cursorPosition,
    doc
  );
  const treeCursor = tree.cursorAt(cursorPosition, -1);
  if (!currentStatement) {
    if (treeCursor.node.name === "LogicalOperator") {
      currentToken = convertedTree?.children.find(
        (token) => isFilterFieldValueNode(token) && token.range.from <= cursorPosition && token.range.to >= cursorPosition
      ) ?? fallbackToken;
    }
  } else {
    for (const [key, value] of Object.entries(currentStatement)) {
      if (isFilterFieldValueNode(value) && value.range.from <= cursorPosition && value.range.to >= cursorPosition) {
        currentToken = value;
        currentTokenType = key;
        break;
      }
    }
  }
  if (Array.isArray(currentToken.value) && isFilterFieldListNode(currentToken)) {
    currentToken = currentToken.value?.find(
      (token) => isFilterFieldLeafNode(token) && token.range.from <= cursorPosition && token.range.to >= cursorPosition
    ) ?? currentToken;
    if (currentToken.type === "List" || currentToken.type === "Error") {
      let nodeCursor = tree.cursorAt(cursorPosition, 1);
      if (nodeCursor.node.name === "Filter") {
        nodeCursor = tree.cursorAt(cursorPosition, -1);
      }
      const { node } = nodeCursor.node;
      const tokensInsideList = [
        "List",
        "Identifier",
        "String",
        "Boolean",
        "Number",
        "Duration"
      ];
      if (tokensInsideList.includes(node.name) || node.name === "OpeningBracket" && cursorPosition > node.from || node.name === "ClosingBracket" && cursorPosition < node.to) {
        currentToken = fallbackToken;
      }
    }
  }
  return { currentToken, currentTokenType };
}
export {
  getCurrentToken
};
