// @dynatrace/strato-components/esm/filters/filter-field/utils/suggestions-utils/key-suggestions.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { FormattedMessage } from "react-intl";
import {
  AbcIcon,
  BooleanIcon,
  ClockIcon,
  HashtagIcon,
  IPaddressIcon,
  JsonIcon,
  StopwatchIcon,
  UIDIcon
} from "@dynatrace/strato-icons";
import {
  getFallbackKeySuggestion,
  getSearchSuggestion,
  getStartTypingSuggestion,
  getStatementSuggestionForMatchingValue
} from "./shared-suggestions.js";
import { getValidValueTypes } from "./suggestion-type.js";
import { filterFieldMessages } from "../../messages.js";
import { KEY_GROUP_ID } from "../validation-utils/normalize-validator-map.js";
import {
  isFilterFieldValidatorMapTypeSpecification,
  isFilterFieldValueSuggestionGroupConfig
} from "../validation-utils/type-guards.js";
import { getValuePredicateTextValue } from "../validation-utils/validation-helper-functions.js";
const JSONPATH_PATTERN_PARTIAL = /^(?:"(?:[^"\\]|\\.)*"|[^\s"$]+)\$(?:(?:\.(?:[a-zA-Z_][\w]*|"(?:[^"\\]|\\.)*")|\[(?:\d+|"(?:[^"\\]|\\.)*")\])*)/;
const SEARCH_OPERATOR_VALUES = ["", "*", "* ", "*~", "* ~"];
const isSearchOperatorValue = (value) => SEARCH_OPERATOR_VALUES.includes(value);
function getKeySuggestions(internalValidatorMap, suggestionTypes = [], customTypes, doc, currentToken) {
  const validatorMapIsExhaustive = internalValidatorMap?.exhaustive;
  let textValue = currentToken?.textValue ?? "";
  const hasTrailingSpace = textValue.endsWith(" ");
  const textValueNeedsEscaping = textValue.toLocaleLowerCase() === "or" || textValue.toLocaleLowerCase() === "and";
  if (textValue && textValueNeedsEscaping && doc && currentToken) {
    textValue = doc.slice(currentToken.range.from, currentToken.range.to).toString();
  }
  if (!internalValidatorMap) {
    if (textValue && !validatorMapIsExhaustive && !hasTrailingSpace && !isSearchOperatorValue(textValue)) {
      return [
        {
          id: KEY_GROUP_ID,
          suggestions: [
            getDefaultKeySuggestion(textValue, textValueNeedsEscaping)
          ]
        }
      ];
    } else if (!textValue && !validatorMapIsExhaustive) {
      return [
        { id: KEY_GROUP_ID, suggestions: [getStartTypingSuggestion("key")] }
      ];
    }
    return [];
  }
  const { byKey, keyToGroup } = internalValidatorMap.normalizedKeys;
  const groupEntryById = /* @__PURE__ */ new Map();
  const result = [];
  let suggestionMatchesTextValue = false;
  for (const [key, config] of byKey) {
    if (key === textValue) {
      suggestionMatchesTextValue = true;
    }
    let suggestionTypeResult;
    if (config.valueType) {
      suggestionTypeResult = getValidValueTypes({ type: config.valueType });
    } else if (config.valuePredicate) {
      suggestionTypeResult = getValidValueTypes(config.valuePredicate);
    } else {
      suggestionTypeResult = {
        suggestionType: "Any",
        hasLiteralValues: false,
        types: ["Any"]
      };
    }
    const { suggestionType, hasLiteralValues } = suggestionTypeResult;
    const displayValue = config.displayValue;
    const details = config.details;
    let keySuggestion;
    if (suggestionType === "JSONPath") {
      keySuggestion = getJSONPathKeySuggestion(key, suggestionType);
    } else if (Array.isArray(suggestionType)) {
      keySuggestion = suggestionType.map((type) => {
        if (type === "JSONPath") {
          return getJSONPathKeySuggestion(key, type);
        }
        return {
          value: key,
          icon: getIconForValueType(type),
          suggestionType: "validatorMap",
          displayValue,
          details
        };
      });
    } else if (hasLiteralValues && Array.isArray(config.valuePredicate)) {
      keySuggestion = {
        value: key,
        icon: getIconForValueType("Any"),
        suggestionType: "validatorMap",
        displayValue,
        details
      };
    } else {
      keySuggestion = {
        value: key,
        icon: getIconForValueType(suggestionType, customTypes),
        suggestionType: "validatorMap",
        displayValue,
        details
      };
    }
    const keySuggestions = Array.isArray(keySuggestion) ? keySuggestion : [keySuggestion];
    const group = keyToGroup.get(key);
    if (group === void 0) {
      continue;
    }
    let groupEntry = groupEntryById.get(group.id);
    if (!groupEntry) {
      groupEntry = {
        id: group.id,
        label: group.label,
        "aria-label": group["aria-label"],
        suggestions: []
      };
      groupEntryById.set(group.id, groupEntry);
      result.push(groupEntry);
    }
    groupEntry.suggestions.push(...keySuggestions);
  }
  const isJSONPathContinuation = currentToken?.type === "JSONPath" || currentToken?.type === "Error" && JSONPATH_PATTERN_PARTIAL.test(currentToken?.textValue ?? "");
  if (textValue && !validatorMapIsExhaustive && !suggestionMatchesTextValue && !isSearchOperatorValue(textValue) && (!hasTrailingSpace || suggestionTypes.includes("pastedContent")) && !isJSONPathContinuation) {
    const trailingSuggestion = getDefaultKeySuggestion(
      textValue,
      textValueNeedsEscaping
    );
    const lastGroup = result.at(-1);
    if (lastGroup?.id.startsWith(`${KEY_GROUP_ID}-implicit-`)) {
      lastGroup.suggestions.push(trailingSuggestion);
    } else {
      result.push({
        id: `${KEY_GROUP_ID}-implicit-${result.length}`,
        suggestions: [trailingSuggestion]
      });
    }
  }
  return result;
}
function getSearchSuggestions(context, currentStatement, currentToken, suggestionTypes, isInsideAGroup) {
  if (isInsideAGroup) {
    return [];
  }
  const textValue = currentToken.textValue;
  const { value } = currentToken;
  if (currentStatement?.comparisonOperator !== void 0 && currentStatement.comparisonOperator.type !== "SearchOperator" || currentStatement?.comparisonOperator?.type === "SearchOperator" && value === "") {
    return [];
  }
  const emptyInput = context.state.doc.toString() === "";
  if (emptyInput || isSearchOperatorValue(textValue)) {
    return [getSearchSuggestion()];
  } else if ((suggestionTypes.includes("key") || suggestionTypes.includes("value") || suggestionTypes.includes("pastedContent")) && !(currentToken.type === "JSONPath" || currentToken.type === "Error" && JSONPATH_PATTERN_PARTIAL.exec(currentToken.textValue) && !currentToken.textValue.endsWith("$"))) {
    return [getSearchSuggestion(textValue)];
  }
  return [];
}
function getFallbackKeySuggestions(currentToken, suggestionTypes, fallbackKeys, filterStatementSuggestions) {
  if (!fallbackKeys || fallbackKeys.length === 0) {
    return [];
  }
  const { value } = currentToken;
  const suggestions = [];
  if (suggestionTypes.includes("key") && value !== "") {
    const textValue = currentToken.textValue;
    for (const key of fallbackKeys) {
      if (Array.isArray(key.comparisonOperators)) {
        for (const operator of key.comparisonOperators) {
          const fallbackKeySuggestion = getFallbackKeySuggestion(
            key.value,
            operator,
            textValue
          );
          if (!(operator === "equals" && filterStatementSuggestions?.some(
            (statementSuggestion) => statementSuggestion.value === fallbackKeySuggestion.value
          ))) {
            suggestions.push(fallbackKeySuggestion);
          }
        }
      } else {
        suggestions.push(
          getFallbackKeySuggestion(key.value, "matches-phrase", textValue)
        );
      }
    }
  }
  return suggestions;
}
function getFilterStatementSuggestions(currentToken, suggestionTypes, filterStatementKeysMap) {
  const { textValue } = currentToken;
  if (!filterStatementKeysMap || !suggestionTypes.includes("key") || textValue === "") {
    return [];
  }
  const suggestions = [];
  if (filterStatementKeysMap.values.size > 0) {
    const lowerTextValue = textValue.toLowerCase();
    for (const [value, keys] of filterStatementKeysMap.values.entries()) {
      if (value.toLowerCase().includes(lowerTextValue)) {
        for (const key of keys) {
          suggestions.push(getStatementSuggestionForMatchingValue(key, value));
        }
      }
    }
  }
  if (filterStatementKeysMap.functions.size > 0) {
    for (const [key, func] of filterStatementKeysMap.functions.entries()) {
      if (func(currentToken)) {
        suggestions.push(
          getStatementSuggestionForMatchingValue(key, currentToken.textValue)
        );
      }
    }
  }
  return suggestions;
}
function getFilterStatementKeysMap(validatorMap) {
  if (!validatorMap || Array.isArray(validatorMap.keyPredicates) || !validatorMap.keyPredicates) {
    return void 0;
  }
  const valuesMap = /* @__PURE__ */ new Map();
  const functionsMap = /* @__PURE__ */ new Map();
  for (const [key, keyPredicate] of Object.entries(
    validatorMap.keyPredicates
  )) {
    if (!isFilterFieldValidatorMapTypeSpecification(keyPredicate)) {
      if (typeof keyPredicate.suggestStatementOnValueMatch === "function") {
        functionsMap.set(key, keyPredicate.suggestStatementOnValueMatch);
      } else if (keyPredicate.suggestStatementOnValueMatch && keyPredicate.valuePredicate) {
        const valuePredicate = keyPredicate.valuePredicate;
        if (Array.isArray(valuePredicate)) {
          for (const value of valuePredicate) {
            if (!isFilterFieldValueSuggestionGroupConfig(value)) {
              addKeyValuePairToMap(valuesMap, key, value);
            }
          }
        } else if (valuePredicate && typeof valuePredicate !== "function" && !isFilterFieldValueSuggestionGroupConfig(valuePredicate)) {
          addKeyValuePairToMap(valuesMap, key, valuePredicate);
        }
      }
    }
  }
  return valuesMap.size > 0 || functionsMap.size > 0 ? { values: valuesMap, functions: functionsMap } : void 0;
}
function addKeyValuePairToMap(valuesMap, key, value) {
  if (!isFilterFieldValidatorMapTypeSpecification(value)) {
    const stringValue = getValuePredicateTextValue(value);
    const existingKeys = valuesMap.get(stringValue) ?? [];
    if (!existingKeys.includes(key)) {
      valuesMap.set(stringValue, [...existingKeys, key]);
    }
  }
}
function getFallbackKeys(validatorMap) {
  if (!validatorMap || Array.isArray(validatorMap.keyPredicates)) {
    return void 0;
  }
  const keys = [];
  for (const key in validatorMap.keyPredicates) {
    const keyPredicate = validatorMap.keyPredicates[key];
    if (!isFilterFieldValidatorMapTypeSpecification(keyPredicate) && keyPredicate.fallbackKey) {
      keys.push({
        value: key,
        comparisonOperators: Array.isArray(keyPredicate.fallbackKey) ? keyPredicate.fallbackKey : void 0
      });
    }
  }
  return keys.length > 0 ? keys : void 0;
}
function getIconForCustomType(type, customTypes) {
  for (const [typeName, customType] of Object.entries(customTypes)) {
    if (typeName === type && typeof customType === "object" && customType.icon !== void 0) {
      return customType.icon;
    }
  }
  return null;
}
function getIconForValueType(type, customTypes) {
  switch (type) {
    case "Number": {
      return /* @__PURE__ */ jsx(HashtagIcon, {});
    }
    case "Boolean": {
      return /* @__PURE__ */ jsx(BooleanIcon, {});
    }
    case "Duration": {
      return /* @__PURE__ */ jsx(StopwatchIcon, {});
    }
    case "JSONPath": {
      return /* @__PURE__ */ jsx(JsonIcon, {});
    }
    case "IPAddress": {
      return /* @__PURE__ */ jsx(IPaddressIcon, {});
    }
    case "UID": {
      return /* @__PURE__ */ jsx(UIDIcon, {});
    }
    case "SmartscapeId": {
      return /* @__PURE__ */ jsx(UIDIcon, {});
    }
    case "Timestamp": {
      return /* @__PURE__ */ jsx(ClockIcon, {});
    }
    case "String":
    case "Any":
      return /* @__PURE__ */ jsx(AbcIcon, {});
  }
  if (customTypes) {
    return getIconForCustomType(type, customTypes);
  }
}
function getDefaultKeySuggestion(value, valueNeedsEscaping = false) {
  const match = /^"(.+)(?<!\\)"$/.exec(value);
  const textValue = match ? match[1] : value;
  return {
    trailing: " ",
    suggestionType: "hint-useAsKey",
    value: valueNeedsEscaping ? `"${value}"` : value,
    id: "strato-filterfield-suggestion-use-as-key",
    displayValue: /* @__PURE__ */ jsx(
      FormattedMessage,
      {
        ...filterFieldMessages.suggestions.useAsKey,
        values: {
          value: /* @__PURE__ */ jsxs(Fragment, { children: [
            "\xA0",
            /* @__PURE__ */ jsx("mark", { children: textValue }),
            "\xA0"
          ] })
        }
      }
    )
  };
}
function getJSONPathKeySuggestion(key, valueType = "Any") {
  return {
    value: `${key}$.`,
    icon: getIconForValueType(valueType),
    trailing: "",
    suggestionType: "JSONPath",
    cursorBeforeTrailing: true,
    details: /* @__PURE__ */ jsx(
      FormattedMessage,
      {
        defaultMessage: "Filter JSON attributes",
        description: "Details shown for JSONPath key suggestions.",
        id: "/SFvkA0WT3AVdLhH"
      }
    ),
    insertionStrategy: "replace-statement"
  };
}
export {
  JSONPATH_PATTERN_PARTIAL,
  getFallbackKeySuggestions,
  getFallbackKeys,
  getFilterStatementKeysMap,
  getFilterStatementSuggestions,
  getKeySuggestions,
  getSearchSuggestions
};
