// @dynatrace/strato-components/esm/filters/filter-field/utils/suggestions-utils/operator-suggestions.js
import { jsx, jsxs } from "react/jsx-runtime";
import { FormattedMessage } from "react-intl";
import Colors from "@dynatrace/strato-design-tokens/colors";
import Typography from "@dynatrace/strato-design-tokens/typography";
import { filterFieldMessages } from "../../messages.js";
import { getValidOperatorsShort } from "../validation-utils/validation-helper-functions.js";
function getComparisonOperatorSuggestions(key, internalValidatorMap, parserConfig) {
  const operators = getValidOperatorsShort(
    key,
    internalValidatorMap,
    parserConfig
  );
  return operators.map((operator) => {
    const operatorWithDetails = operatorSuggestionsWithDetails[operator];
    const baseProps = {
      suggestionType: "validatorMap"
    };
    return operatorWithDetails ? { ...operatorWithDetails, ...baseProps } : { value: operator, ...baseProps };
  });
}
const defaultLogicalOperatorSuggestions = [
  { value: "OR" }
];
const ValuePlaceholder = () => {
  return /* @__PURE__ */ jsx(
    "span",
    {
      style: {
        color: Colors.Text.Neutral.Subdued,
        fontWeight: Typography.Text.Base.Default.Weight
      },
      children: "value"
    }
  );
};
const operatorSuggestionsWithDetails = {
  "=": {
    value: "=",
    details: /* @__PURE__ */ jsx(FormattedMessage, { ...filterFieldMessages.comparisonOperators.equals }),
    escapeSuggestion: false
  },
  "!=": {
    value: "!=",
    details: /* @__PURE__ */ jsx(
      FormattedMessage,
      {
        ...filterFieldMessages.comparisonOperators.notEquals
      }
    ),
    escapeSuggestion: false
  },
  "<": {
    value: "<",
    details: /* @__PURE__ */ jsx(FormattedMessage, { ...filterFieldMessages.comparisonOperators.lessThan }),
    escapeSuggestion: false
  },
  "<=": {
    value: "<=",
    details: /* @__PURE__ */ jsx(
      FormattedMessage,
      {
        ...filterFieldMessages.comparisonOperators.lessThanEqual
      }
    ),
    escapeSuggestion: false
  },
  ">": {
    value: ">",
    details: /* @__PURE__ */ jsx(
      FormattedMessage,
      {
        ...filterFieldMessages.comparisonOperators.greaterThan
      }
    ),
    escapeSuggestion: false
  },
  ">=": {
    value: ">=",
    details: /* @__PURE__ */ jsx(
      FormattedMessage,
      {
        ...filterFieldMessages.comparisonOperators.greaterThanEqual
      }
    ),
    escapeSuggestion: false
  },
  in: {
    value: "in (",
    displayValue: "in ()",
    details: /* @__PURE__ */ jsx(FormattedMessage, { ...filterFieldMessages.comparisonOperators.in }),
    trailing: ") ",
    cursorBeforeTrailing: true,
    escapeSuggestion: false
  },
  "not in": {
    value: "not in (",
    displayValue: "not in ()",
    details: /* @__PURE__ */ jsx(FormattedMessage, { ...filterFieldMessages.comparisonOperators.notIn }),
    trailing: ") ",
    cursorBeforeTrailing: true,
    escapeSuggestion: false
  },
  "starts-with": {
    value: '= "',
    displayValue: /* @__PURE__ */ jsxs("div", { children: [
      "= ",
      /* @__PURE__ */ jsx(ValuePlaceholder, {}),
      "*"
    ] }),
    details: /* @__PURE__ */ jsx(
      FormattedMessage,
      {
        ...filterFieldMessages.comparisonOperators.startsWith
      }
    ),
    trailing: 'value"*',
    cursorBeforeTrailing: true,
    escapeSuggestion: false
  },
  contains: {
    value: '= *"',
    displayValue: /* @__PURE__ */ jsxs("div", { children: [
      "= *",
      /* @__PURE__ */ jsx(ValuePlaceholder, {}),
      "*"
    ] }),
    details: /* @__PURE__ */ jsx(FormattedMessage, { ...filterFieldMessages.comparisonOperators.contains }),
    trailing: 'value"*',
    cursorBeforeTrailing: true,
    escapeSuggestion: false
  },
  "ends-with": {
    value: '= *"',
    displayValue: /* @__PURE__ */ jsxs("div", { children: [
      "= *",
      /* @__PURE__ */ jsx(ValuePlaceholder, {})
    ] }),
    details: /* @__PURE__ */ jsx(FormattedMessage, { ...filterFieldMessages.comparisonOperators.endsWith }),
    trailing: 'value"',
    cursorBeforeTrailing: true,
    escapeSuggestion: false
  },
  "not-starts-with": {
    value: '!= "',
    displayValue: /* @__PURE__ */ jsxs("div", { children: [
      "!= ",
      /* @__PURE__ */ jsx(ValuePlaceholder, {}),
      "*"
    ] }),
    details: /* @__PURE__ */ jsx(
      FormattedMessage,
      {
        ...filterFieldMessages.comparisonOperators.notStartsWith
      }
    ),
    trailing: 'value"*',
    cursorBeforeTrailing: true,
    escapeSuggestion: false
  },
  "not-contains": {
    value: '!= *"',
    displayValue: /* @__PURE__ */ jsxs("div", { children: [
      "!= *",
      /* @__PURE__ */ jsx(ValuePlaceholder, {}),
      "*"
    ] }),
    details: /* @__PURE__ */ jsx(
      FormattedMessage,
      {
        ...filterFieldMessages.comparisonOperators.notContains
      }
    ),
    trailing: 'value"*',
    cursorBeforeTrailing: true,
    escapeSuggestion: false
  },
  "not-ends-with": {
    value: '!= *"',
    displayValue: /* @__PURE__ */ jsxs("div", { children: [
      "!= *",
      /* @__PURE__ */ jsx(ValuePlaceholder, {})
    ] }),
    details: /* @__PURE__ */ jsx(
      FormattedMessage,
      {
        ...filterFieldMessages.comparisonOperators.notEndsWith
      }
    ),
    trailing: 'value"',
    cursorBeforeTrailing: true,
    escapeSuggestion: false
  },
  "= *": {
    value: "= *",
    details: /* @__PURE__ */ jsx(FormattedMessage, { ...filterFieldMessages.comparisonOperators.exists })
  },
  "!= *": {
    value: "!= *",
    details: /* @__PURE__ */ jsx(
      FormattedMessage,
      {
        ...filterFieldMessages.comparisonOperators.notExists
      }
    ),
    escapeSuggestion: false
  },
  "~": {
    value: "~",
    details: /* @__PURE__ */ jsx(
      FormattedMessage,
      {
        ...filterFieldMessages.comparisonOperators.matchesPhrase
      }
    ),
    escapeSuggestion: false
  },
  "!~": {
    value: "!~",
    details: /* @__PURE__ */ jsx(
      FormattedMessage,
      {
        ...filterFieldMessages.comparisonOperators.notMatchesPhrase
      }
    ),
    escapeSuggestion: false
  }
};
export {
  ValuePlaceholder,
  defaultLogicalOperatorSuggestions,
  getComparisonOperatorSuggestions,
  operatorSuggestionsWithDetails
};
