// @dynatrace/strato-components/esm/filters/filter-field/utils/suggestions-utils/shared-suggestions.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { FormattedMessage } from "react-intl";
import { MagnifyingGlassIcon, PasteIcon } from "@dynatrace/strato-icons";
import * as styles from "../../FilterField.sty.js";
import { filterFieldMessages } from "../../messages.js";
import * as suggestionStyles from "../../suggestions/Suggestion.sty.js";
import { escapeSuggestion } from "../escape-suggestion.js";
import { getOperatorShortName } from "../validation-utils/operator-specification.js";
function getNoMatchingSuggestion(valueType = "Any") {
  return {
    trailing: "",
    valueType,
    displayValue: /* @__PURE__ */ jsx(
      FormattedMessage,
      {
        ...filterFieldMessages.suggestions.noMatchingSuggestions
      }
    )
  };
}
function getStartTypingSuggestion(suggestionType, valueType = "Any") {
  return {
    trailing: "",
    valueType,
    suggestionType: "hint-startTyping",
    displayValue: /* @__PURE__ */ jsx(
      FormattedMessage,
      {
        ...filterFieldMessages.suggestions.startTyping,
        values: {
          suggestionType
        }
      }
    )
  };
}
function getPastingSuggestion(valueType = "Any") {
  return {
    trailing: "",
    valueType,
    suggestionType: "pastedContent",
    id: "strato-filterfield-suggestion-pasted-content",
    icon: /* @__PURE__ */ jsx(PasteIcon, {}),
    displayValue: /* @__PURE__ */ jsx(FormattedMessage, { ...filterFieldMessages.suggestions.pasteRaw }),
    "data-strato-filter-field-should-escape-suggestion": true
  };
}
function getSearchSuggestion(value) {
  if (value === void 0) {
    return {
      value: '* ~ "',
      trailing: '"',
      cursorBeforeTrailing: true,
      id: "strato-filterfield-suggestion-search",
      displayValue: /* @__PURE__ */ jsx(
        FormattedMessage,
        {
          ...filterFieldMessages.suggestions.search,
          values: {
            value: false
          }
        }
      ),
      icon: /* @__PURE__ */ jsx(MagnifyingGlassIcon, {}),
      insertionStrategy: "replace-statement"
    };
  }
  const match = /^"(.+)(?<!\\)" ?$/.exec(value);
  const textValue = match ? match[1] : value;
  const escapedSuggestion = escapeSuggestion(match ? value.trim() : value);
  return {
    value: `* ~ ${escapedSuggestion}`,
    id: "strato-filterfield-suggestion-search",
    displayValue: /* @__PURE__ */ jsx(
      FormattedMessage,
      {
        ...filterFieldMessages.suggestions.search,
        values: {
          value: /* @__PURE__ */ jsxs(Fragment, { children: [
            "\xA0",
            /* @__PURE__ */ jsx("mark", { children: textValue })
          ] })
        }
      }
    ),
    icon: /* @__PURE__ */ jsx(MagnifyingGlassIcon, {}),
    insertionStrategy: "replace-statement"
  };
}
function getFallbackKeySuggestion(key, operator, value) {
  const match = /^"(.+)(?<!\\)" ?$/.exec(value);
  const textValue = match ? match[1] : value;
  const escapedSuggestion = escapeSuggestion(match ? value.trim() : value);
  const baseOperator = operator.includes("not") ? "!=" : "=";
  switch (operator) {
    case "in":
    case "not in": {
      return {
        value: `${key} ${operator} (${escapedSuggestion})`,
        displayValue: /* @__PURE__ */ jsxs("span", { children: [
          key,
          " ",
          operator,
          " (",
          /* @__PURE__ */ jsx("mark", { children: textValue }),
          ")"
        ] }),
        insertionStrategy: "replace-statement"
      };
    }
    case "contains":
    case "not-contains": {
      return {
        value: `${key} ${baseOperator} *${escapedSuggestion}*`,
        displayValue: /* @__PURE__ */ jsxs("span", { children: [
          key,
          " ",
          baseOperator,
          " *",
          /* @__PURE__ */ jsx("mark", { children: textValue }),
          "*"
        ] }),
        insertionStrategy: "replace-statement"
      };
    }
    case "starts-with":
    case "not-starts-with": {
      return {
        value: `${key} ${baseOperator} ${escapedSuggestion}*`,
        displayValue: /* @__PURE__ */ jsxs("span", { children: [
          key,
          " ",
          baseOperator,
          " ",
          /* @__PURE__ */ jsx("mark", { children: textValue }),
          "*"
        ] }),
        insertionStrategy: "replace-statement"
      };
    }
    case "ends-with":
    case "not-ends-with": {
      return {
        value: `${key} ${baseOperator} *${escapedSuggestion}`,
        displayValue: /* @__PURE__ */ jsxs("span", { children: [
          key,
          " ",
          baseOperator,
          " *",
          /* @__PURE__ */ jsx("mark", { children: textValue })
        ] }),
        insertionStrategy: "replace-statement"
      };
    }
    default: {
      const operatorName = getOperatorShortName(operator);
      return {
        value: `${key} ${operatorName} ${escapedSuggestion}`,
        displayValue: /* @__PURE__ */ jsxs("span", { children: [
          key,
          " ",
          operatorName,
          " ",
          /* @__PURE__ */ jsx("mark", { children: textValue })
        ] }),
        insertionStrategy: "replace-statement"
      };
    }
  }
}
function getStatementSuggestionForMatchingValue(key, value) {
  const match = /^"(.+)(?<!\\)" ?$/.exec(value);
  const textValue = match ? match[1] : value;
  const escapedSuggestion = escapeSuggestion(match ? value.trim() : value);
  return {
    value: `${key} = ${escapedSuggestion}`,
    // Add the class only to the value part as we only want to show highlighting in the value and not the key
    displayValue: /* @__PURE__ */ jsxs("span", { children: [
      key,
      " =",
      " ",
      /* @__PURE__ */ jsx("span", { className: suggestionStyles.statementSuggestionValue, children: textValue })
    ] }),
    insertionStrategy: "replace-statement"
  };
}
function getJSONPathHint() {
  return {
    trailing: "",
    id: "strato-filterfield-suggestion-jsonpath-hint",
    suggestionType: "JSONPath",
    displayValue: /* @__PURE__ */ jsxs("span", { className: styles.suggestionHintContainer, children: [
      /* @__PURE__ */ jsx(
        FormattedMessage,
        {
          defaultMessage: "Filter JSON attributes using dot notation. For example: ",
          description: "Hint for entering JSONPath syntax in the FilterField.",
          id: "Ua7PnIRWAcErpii9"
        }
      ),
      /* @__PURE__ */ jsxs("span", { className: styles.filterStatementBase, children: [
        /* @__PURE__ */ jsx("span", { className: styles.filterKey, children: "content$.key[1].key " }),
        /* @__PURE__ */ jsx("span", { className: styles.operator, children: "=" }),
        " ",
        /* @__PURE__ */ jsx("span", { className: styles.filterValue, children: "value" })
      ] })
    ] })
  };
}
export {
  getFallbackKeySuggestion,
  getJSONPathHint,
  getNoMatchingSuggestion,
  getPastingSuggestion,
  getSearchSuggestion,
  getStartTypingSuggestion,
  getStatementSuggestionForMatchingValue
};
