// @dynatrace/strato-components/esm/filters/filter-field/utils/suggestions-utils/suggestion-type.js
import {
  isFilterFieldRichValuePredicate,
  isFilterFieldValidatorMapDurationSpecification,
  isFilterFieldValidatorMapTypeSpecification,
  isFilterFieldValueSuggestionGroupConfig
} from "../validation-utils/type-guards.js";
import {
  isValidIPAddress,
  isValidSmartscapeId,
  isValidTimestamp,
  isValidUID
} from "../validation-utils/type-validators.js";
function getValidValueTypes(valuePredicate) {
  let valueTypes;
  let hasLiteralValues = false;
  if (typeof valuePredicate === "function") {
    valueTypes = ["Any"];
  } else {
    ({ types: valueTypes, hasLiteralValues } = getTypesFromValuePredicate(valuePredicate));
  }
  valueTypes = Array.from(new Set(valueTypes));
  const suggestionType = getSuggestionTypesFromPrimitiveArray(valueTypes);
  return { types: valueTypes, suggestionType, hasLiteralValues };
}
function getSuggestionTypeForStringSuggestions(valuePredicate) {
  switch (true) {
    case isValidIPAddress(valuePredicate):
      return { types: ["IPAddress"], hasLiteralValues: true };
    case isValidTimestamp(valuePredicate):
      return { types: ["Timestamp"], hasLiteralValues: true };
    case isValidUID(valuePredicate):
      return { types: ["UID"], hasLiteralValues: true };
    case isValidSmartscapeId(valuePredicate):
      return { types: ["SmartscapeId"], hasLiteralValues: true };
    default:
      return { types: ["String"], hasLiteralValues: true };
  }
}
function getTypesFromValuePredicate(valuePredicate) {
  if (valuePredicate === void 0) {
    return { types: ["Any"], hasLiteralValues: false };
  }
  if (isFilterFieldRichValuePredicate(valuePredicate)) {
    return getTypesFromValuePredicate(valuePredicate.value);
  }
  if (typeof valuePredicate === "string") {
    return getSuggestionTypeForStringSuggestions(valuePredicate);
  }
  if (typeof valuePredicate === "number") {
    return { types: ["Number"], hasLiteralValues: true };
  }
  if (typeof valuePredicate === "boolean") {
    return { types: ["Boolean"], hasLiteralValues: true };
  }
  if (isFilterFieldValidatorMapDurationSpecification(valuePredicate)) {
    return { types: ["Duration"], hasLiteralValues: true };
  }
  if (isFilterFieldValidatorMapTypeSpecification(valuePredicate)) {
    const predicate = valuePredicate.type;
    if (Array.isArray(predicate)) {
      return {
        types: predicate.length > 0 ? Array.from(new Set(predicate)) : ["Any"],
        hasLiteralValues: false
      };
    }
    return { types: [predicate], hasLiteralValues: false };
  }
  if (isFilterFieldValueSuggestionGroupConfig(valuePredicate)) {
    return { types: ["Any"], hasLiteralValues: true };
  }
  return getTypesFromPredicateArray(valuePredicate);
}
function getTypesFromPredicateArray(valuePredicate) {
  const types = [];
  let hasLiteralValues = false;
  for (const predicate of valuePredicate) {
    if (isFilterFieldValidatorMapTypeSpecification(predicate)) {
      if (Array.isArray(predicate.type)) {
        types.push(...predicate.type);
      } else {
        types.push(predicate.type);
      }
    } else {
      hasLiteralValues = true;
    }
  }
  return {
    types: types.length > 0 ? Array.from(new Set(types)) : ["Any"],
    hasLiteralValues
  };
}
function getSuggestionTypesFromPrimitiveArray(primitives) {
  if (primitives.length === 1) {
    return primitives[0];
  }
  if (primitives.includes("JSONPath")) {
    if (primitives.length <= 2) {
      return primitives.sort((a, b) => a === "JSONPath" ? 1 : -1);
    }
    return ["Any", "JSONPath"];
  }
  return "Any";
}
export {
  getValidValueTypes
};
