// @dynatrace/strato-components/charts/core/components/single-value-renderer/utils/build-value-content-grid-rows.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var build_value_content_grid_rows_exports = {};
__export(build_value_content_grid_rows_exports, {
  buildValueContentGridRows: () => buildValueContentGridRows,
  componentsHeightMapToCSSGridTemplateString: () => componentsHeightMapToCSSGridTemplateString,
  getInnerGridValue: () => getInnerGridValue,
  sumFractions: () => sumFractions
});
module.exports = __toCommonJS(build_value_content_grid_rows_exports);
const fractions = {
  label: 3,
  value: 6,
  trend: 3,
  sparkline: 8
};
const sumFractions = (configWithFractions) => configWithFractions.reduce((acc, cur) => acc + cur, 0);
const componentsHeightMapToCSSGridTemplateString = (fractionsMap) => Array.from(fractionsMap).map(([key, value]) => `"${key}" ${value}`).join("\r\n");
const getInnerGridValue = (innerContentMap, localFractionSum) => {
  const innerGridPercentValues = /* @__PURE__ */ new Map();
  let allAreasHeight = 0;
  innerContentMap.forEach((value, key) => {
    const areaHeight = Math.round(value / localFractionSum * 100);
    allAreasHeight += areaHeight;
    const isRoundedOnePercentLeft = 100 - allAreasHeight === 1;
    innerGridPercentValues.set(
      key,
      `${isRoundedOnePercentLeft ? areaHeight + 1 : areaHeight}cqh`
    );
  });
  return componentsHeightMapToCSSGridTemplateString(innerGridPercentValues);
};
const getOuterGridValue = (outerContentMap, fractionsSum) => {
  const outerGridPercentValues = /* @__PURE__ */ new Map();
  outerContentMap.forEach(
    (value, key) => outerGridPercentValues.set(
      key,
      `${Math.round(value / fractionsSum * 100)}cqh`
    )
  );
  const valueHeight = 100 - parseInt(outerGridPercentValues.get("label") ?? "0");
  outerGridPercentValues.set("content", `${Math.round(valueHeight)}cqh`);
  return componentsHeightMapToCSSGridTemplateString(outerGridPercentValues);
};
function buildValueContentGridRows(hasLabel, hasValue, hasTrend, hasSparkline) {
  const innerComponents = /* @__PURE__ */ new Map();
  const outerComponents = /* @__PURE__ */ new Map();
  hasValue && innerComponents.set("value", fractions.value);
  hasTrend && innerComponents.set("trend", fractions.trend);
  hasSparkline && innerComponents.set("sparkline", fractions.sparkline);
  hasLabel && outerComponents.set("label", fractions.label);
  const innerComponentsFractionsArray = [...innerComponents.values()];
  const innerFractionSum = sumFractions(innerComponentsFractionsArray);
  const outerFractionSum = sumFractions([
    ...innerComponentsFractionsArray,
    ...outerComponents.values()
  ]);
  return {
    innerGrid: getInnerGridValue(innerComponents, innerFractionSum),
    outerGrid: getOuterGridValue(outerComponents, outerFractionSum)
  };
}
