// @dynatrace/strato-components/esm/charts/categorical-bar/hooks/useGenerateCategoricalBarData.js
import { useCategoricalBarChartBaseConfig } from "./useCategoricalBarChartBaseConfig.js";
import { useColorRule } from "../../core/color-rule/useColorRule.js";
import { useColorRuleConfig } from "../../core/context/ColorRuleContext.js";
import {
  isCustomColorPalette
} from "../../core/types/color-palette.js";
import { isValidColorPalette } from "../../core/utils/is-valid-color-palette.js";
import {
  handleMultipleDimensions,
  handleSingleDimension
} from "../utils/generate-categorical-bar-data-helpers.js";
import { isGrouped } from "../utils/layout.js";
function buildSeriesForColorRule(initialData) {
  const series = [];
  initialData.forEach((categoryData) => {
    if (typeof categoryData.value === "object" && categoryData.value !== null) {
      Object.entries(categoryData.value).forEach(
        ([dimensionKey, dimensionValue]) => {
          series.push({
            ...categoryData,
            value: { [dimensionKey]: dimensionValue },
            _originalData: categoryData,
            _dimensionKey: dimensionKey
          });
        }
      );
    } else {
      series.push({
        ...categoryData,
        _originalData: categoryData
      });
    }
  });
  return series;
}
function createCategoryMetadata(id, name) {
  return { id, name };
}
function useGenerateCategoricalBarData(initialData) {
  const {
    valueRepresentation,
    colorPalette: configuredColorPalette,
    colorPaletteMode,
    groupMode
  } = useCategoricalBarChartBaseConfig();
  const isRelative = valueRepresentation === "relative";
  const firstCategoryUnit = initialData[0]?.unit;
  const colorRuleConfig = useColorRuleConfig();
  const hasColorRules = colorRuleConfig.length > 0;
  const seriesForColorRule = hasColorRules ? buildSeriesForColorRule(initialData) : [];
  const colorRules = useColorRule(seriesForColorRule, "categoricalBar");
  const uniqueDimensions = /* @__PURE__ */ new Set();
  const categoricalData = [];
  const categoryMap = /* @__PURE__ */ new WeakMap();
  const dimensionMap = /* @__PURE__ */ new WeakMap();
  const colorByDimension = /* @__PURE__ */ new Map();
  const colorByCategory = /* @__PURE__ */ new Map();
  for (const categoryData of initialData) {
    const categoryId = crypto.randomUUID();
    const {
      unit,
      category: categoryName,
      value: categoryValue,
      color: categoryColor,
      ...rest
    } = categoryData;
    const newCategory = {
      id: categoryId,
      name: categoryName,
      unit,
      dimensions: [],
      ...rest
    };
    const isSingleDimension = unit !== firstCategoryUnit || typeof categoryValue !== "object";
    const categoricalColorPalette = colorPaletteMode === "multi-color" || isGrouped(groupMode) || !isSingleDimension;
    const colorPalette = isValidColorPalette(configuredColorPalette) || isCustomColorPalette(configuredColorPalette) ? configuredColorPalette : void 0;
    if (isSingleDimension) {
      const colorPositionIndex = initialData.indexOf(categoryData);
      const paletteColorPosition = categoricalColorPalette ? colorPositionIndex : 0;
      handleSingleDimension({
        categoryData,
        rest,
        categoryName,
        categoryColor,
        colorByDimension,
        colorRules,
        colorPalette,
        colorPaletteMode,
        paletteColorPosition,
        uniqueDimensions,
        dimensionMap,
        newCategory,
        seriesForColorRule,
        hasColorRules
      });
    } else {
      handleMultipleDimensions({
        categoryData,
        rest,
        categoryName,
        colorByDimension,
        colorByCategory,
        colorRules,
        colorPalette,
        categoricalColorPalette,
        uniqueDimensions,
        dimensionMap,
        newCategory,
        seriesForColorRule,
        hasColorRules
      });
    }
    const categoryMedatada = createCategoryMetadata(categoryId, categoryName);
    categoryMap.set(newCategory, categoryMedatada);
    categoricalData.push(newCategory);
  }
  return {
    categoryMap,
    dimensionMap,
    categoricalData,
    uniqueDimensions: Array.from(uniqueDimensions),
    unit: isRelative ? "percent" : firstCategoryUnit,
    valueRepresentation
  };
}
export {
  useGenerateCategoricalBarData
};
