// @dynatrace/strato-components/esm/charts/categorical-bar/hooks/useGenerateFilteredCategoricalBarData.js
import { useCategoricalBarChartBaseConfig } from "./useCategoricalBarChartBaseConfig.js";
import { getRelativeValue } from "../../core/utils/get-relative-value.js";
import { safeGet } from "../../core/utils/safe-get.js";
import { isStacked } from "../utils/layout.js";
const getMaxDimensionValue = (filteredCategories) => Math.max(
  ...filteredCategories.map(
    (category) => category.dimensions.length > 0 ? Math.abs(category.dimensions[0].value) : 0
  )
);
const calculateTotalCategoryValue = (categoryData) => categoryData.dimensions.reduce(
  (total, dimension) => total + Math.abs(dimension.value),
  0
);
const calculateRelativeValue = (value, relativeTotalValue) => {
  const numericValue = Number(value);
  const numberValue = isNaN(numericValue) ? 0 : numericValue;
  return relativeTotalValue ? getRelativeValue(numberValue, relativeTotalValue) : numberValue;
};
function useGenerateFilteredCategoricalBarData(filteredCategories, categoricalData, unit, categoryMap, dimensionMap, valueRepresentation) {
  const { groupMode } = useCategoricalBarChartBaseConfig();
  const filteredUniqueDimensions = /* @__PURE__ */ new Set();
  const minMaxPerUnit = /* @__PURE__ */ new Map();
  let isSingleDimension = true;
  for (const filteredCategory of filteredCategories) {
    const { id: categoryId, dimensions: filteredDimensions } = filteredCategory;
    if (filteredDimensions.length > 1) {
      isSingleDimension = false;
    }
    const categoryObjReference = categoricalData.find(
      (p) => p.id === categoryId
    );
    const categoryMetadata = safeGet(categoryMap, categoryObjReference);
    let relativeTotalValue = void 0;
    if (valueRepresentation === "relative") {
      relativeTotalValue = isSingleDimension ? getMaxDimensionValue(filteredCategories) : calculateTotalCategoryValue(filteredCategory);
    }
    const cumulativeRanges = [0, 0];
    for (const dimension of filteredDimensions) {
      const dimensionObjReference = categoryObjReference?.dimensions.find(
        (d) => d.name === dimension.name
      );
      const dimensionMetadata = safeGet(dimensionMap, dimensionObjReference);
      let { value } = dimensionMetadata;
      if (valueRepresentation === "relative") {
        value = calculateRelativeValue(dimension.value, relativeTotalValue);
      }
      let cumulativeValue = 0;
      if (isStacked(groupMode)) {
        const NEGATIVE_RANGE = 0;
        const POSITIVE_RANGE = 1;
        const rangeIndex = value < 0 ? NEGATIVE_RANGE : POSITIVE_RANGE;
        cumulativeValue = cumulativeRanges[rangeIndex];
        cumulativeRanges[rangeIndex] += value;
      }
      filteredUniqueDimensions.add(dimension.name);
      dimensionMetadata.value = value;
      dimensionMetadata.cumulativeValue = cumulativeValue;
      dimensionMap.set(dimension, dimensionMetadata);
      const valueToMinMax = (cumulativeValue ?? 0) + value;
      const currentMinMax = minMaxPerUnit.get(unit);
      if (!currentMinMax) {
        minMaxPerUnit.set(unit, {
          min: valueToMinMax,
          max: valueToMinMax
        });
      } else {
        minMaxPerUnit.set(unit, {
          min: Math.min(currentMinMax.min, valueToMinMax),
          max: Math.max(currentMinMax.max, valueToMinMax)
        });
      }
    }
    categoryMap.set(filteredCategory, categoryMetadata);
  }
  return {
    categoryMap,
    dimensionMap,
    filteredUniqueDimensions: Array.from(filteredUniqueDimensions),
    minMaxPerUnit,
    isSingleDimension
  };
}
export {
  calculateRelativeValue,
  calculateTotalCategoryValue,
  getMaxDimensionValue,
  useGenerateFilteredCategoricalBarData
};
