// @dynatrace/strato-components/filters/filter-field/widgets/DeleteStatementWidget.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DeleteStatementWidget_exports = {};
__export(DeleteStatementWidget_exports, {
  DeleteStatementWidget: () => DeleteStatementWidget
});
module.exports = __toCommonJS(DeleteStatementWidget_exports);
var import_view = require("@codemirror/view");
var styles = __toESM(require("../FilterField.sty.js"));
class DeleteStatementWidget extends import_view.WidgetType {
  from;
  to;
  buttonStyles;
  constructor(from, to, incomplete) {
    super();
    this.from = from;
    this.to = to;
    this.buttonStyles = styles.filterStatementDeleteButton({
      incomplete
    });
  }
  toDOM() {
    const wrapper = document.createElement("span");
    wrapper.setAttribute("aria-hidden", "true");
    wrapper.classList.add(styles.filterStatementDeleteWidget);
    const range = document.createRange();
    const deleteButton = range.createContextualFragment(
      // Don't add spaces/linebreaks to the string since they will end up in the markup as well.
      `<button type="button" class="${this.buttonStyles}" tabindex="-1" aria-label="Delete filter statement" data-dt-name="Remove" data-from="${this.from}" data-to="${this.to}"><svg class="${styles.filterStatementDeleteIcon}" xmlns="http://www.w3.org/2000/svg" width="16" height="16" viewBox="0 0 16 16" fill="currentColor"><path d="M8.01524 8.76946L11.9663 12.5459L12.8337 11.7167L8.88272 7.94032L12.7085 4.28358L11.841 3.45443L8.01524 7.11117L4.15895 3.42529L3.29148 4.25444L7.14776 7.94032L3.16626 11.7459L4.03374 12.575L8.01524 8.76946Z" fill="currentColor"/></svg></button>`
    );
    wrapper.appendChild(deleteButton);
    return wrapper;
  }
  ignoreEvent() {
    return false;
  }
}
