// @dynatrace/strato-components/charts/tree-map/components/TreeMapTooltip/TreeMapOverlayTooltip.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TreeMapOverlayTooltip_exports = {};
__export(TreeMapOverlayTooltip_exports, {
  TreeMapOverlayTooltip: () => TreeMapOverlayTooltip
});
module.exports = __toCommonJS(TreeMapOverlayTooltip_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_utils = require("../../../core/components/overlay-tooltip/_to-be-removed/utils.js");
var import_OverlayTooltip = require("../../../core/components/overlay-tooltip/OverlayTooltip.js");
var import_utils2 = require("../../../core/components/overlay-tooltip/utils.js");
var import_useTreeMapCustomTooltip = require("../../hooks/useTreeMapCustomTooltip.js");
var import_useTruncationMode = require("../../hooks/useTruncationMode.js");
const TreeMapOverlayTooltip = (props) => {
  const data = (0, import_utils.useTooltipData)().data;
  const truncationMode = (0, import_useTruncationMode.useTruncationMode)();
  const customTooltip = (0, import_useTreeMapCustomTooltip.useTreeMapCustomTooltip)();
  if (customTooltip?.hidden || (0, import_utils2.isTooltipDataEmpty)(data)) {
    return null;
  }
  const series = data.series[0];
  const meta = data.metadata;
  if (customTooltip?.template) {
    const payload = {
      hierarchyPath: meta.path ?? [series.name],
      value: meta.value,
      childrenCount: meta.children
    };
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayTooltip.OverlayTooltip, { legacyTemplate: customTooltip.template(payload) });
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_OverlayTooltip.OverlayTooltip, { truncation: truncationMode, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayTooltip.OverlayTooltip.Header, { header1: data.header1 }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_OverlayTooltip.OverlayTooltip.SeriesItem,
      {
        name: series.name,
        value: series.value,
        color: series.color,
        actions: props.actionsForItem?.(series)
      }
    )
  ] });
};
TreeMapOverlayTooltip["displayName"] = "TreeMapOverlayTooltip";
