// @dynatrace/strato-components/esm/charts/core/components/interactions-overlay/utils/get-pan-interactions.js
import {
  panDragEnd,
  panDragEndWithMiddleButton,
  panDragMove,
  panDragMoveWithMiddleButton,
  panDragStart,
  panDragStartWithMiddleButton
} from "../interactions/pan/dragInteractions.js";
import {
  panKeyboardEscape,
  panKeyboardLeft,
  panKeyboardRight
} from "../interactions/pan/keyboardInteractions.js";
import {
  panMouseClickOrTap,
  panMouseHover,
  panMouseWheel
} from "../interactions/pan/mouseInteractions.js";
import {
  zoomKeyboardDown,
  zoomKeyboardUp
} from "../interactions/zoom/keyboardInteractions.js";
import { reset } from "./reset.js";
const getPanInteractions = ({
  panDeltaRef,
  onPanStart,
  selectionArea,
  eventCancelled,
  onSelectEnd,
  lastMousePositionRef,
  pinned,
  onHover,
  onTogglePin,
  setCursorMode,
  height,
  onReset,
  width,
  onZoomChange,
  onModeChange,
  onPanEnd,
  onPanChange,
  cursorMode
}) => {
  return {
    onMouseHover: (event) => panMouseHover({
      event,
      pinned,
      lastMousePositionRef,
      cursorMode,
      setCursorMode,
      onHover
    }),
    onMouseClickOrTap: (event) => panMouseClickOrTap({ event, lastMousePositionRef }),
    onDragStart: (event) => panDragStart({ panDeltaRef, onPanStart }),
    onDragMove: (event) => panDragMove({ event, height, panDeltaRef, onPanChange }),
    onDragEnd: (event) => panDragEnd({ panDeltaRef, onPanEnd }),
    onDragStartWithMiddleButton: () => panDragStartWithMiddleButton({ panDeltaRef, onPanStart }),
    onDragMoveWithMiddleButton: (event) => panDragMoveWithMiddleButton({ event, panDeltaRef, onPanChange }),
    onDragEndWithMiddleButton: (event) => panDragEndWithMiddleButton({
      event,
      panDeltaRef,
      lastMousePositionRef,
      onPanEnd
    }),
    onDoubleClickOrTap: (triggeredBy) => {
      onReset?.({
        triggeredBy
      });
      reset(lastMousePositionRef, panDeltaRef);
    },
    onWheel: (event) => panMouseWheel({ event, onPanChange }),
    onKeyEscape: () => panKeyboardEscape({
      selectionArea,
      eventCancelled,
      onSelectEnd,
      lastMousePositionRef,
      pinned,
      onHover,
      onTogglePin,
      onModeChange,
      onPanEnd
    }),
    onKeyEnter: () => {
      return;
    },
    onKeyRight: (shiftKey) => panKeyboardRight({ shiftKey, width, onPanChange }),
    onKeyLeft: (shiftKey) => panKeyboardLeft({ shiftKey, width, onPanChange }),
    onKeyUp: (isTemporaryZoomMode) => {
      if (isTemporaryZoomMode) {
        zoomKeyboardUp({ width, onZoomChange });
      }
    },
    onKeyDown: (isTemporaryZoomMode) => {
      if (isTemporaryZoomMode) {
        zoomKeyboardDown({ width, onZoomChange });
      }
    }
  };
};
export {
  getPanInteractions
};
