// @dynatrace/strato-components/esm/charts/core/components/interactions-overlay/utils/get-interface-interactions.js
import { getActiveMode } from "./get-active-mode.js";
const getInterfaceInteractions = (interactions, mode, isZoomDisabled, isZoomXDisabled, isPanDisabled) => ({
  onDragStart: (event) => interactions[getActiveMode(mode, event, isZoomXDisabled)].onDragStart(
    event
  ),
  onDragMove: (event) => interactions[getActiveMode(mode, event, isZoomXDisabled)].onDragMove(event),
  onDragEnd: (event) => interactions[getActiveMode(mode, event, isZoomXDisabled)].onDragEnd(event),
  onDragStartWithMiddleButton: () => interactions[getActiveMode(mode)].onDragStartWithMiddleButton(),
  onDragMoveWithMiddleButton: (event) => interactions[getActiveMode(mode, event, isZoomXDisabled)].onDragMoveWithMiddleButton(event),
  onDragEndWithMiddleButton: (event) => interactions[getActiveMode(mode, event, isZoomXDisabled)].onDragEndWithMiddleButton(event),
  onTap: (event) => interactions[getActiveMode(mode, event)].onMouseClickOrTap(event),
  onDoubleTap: () => interactions[getActiveMode(mode)].onDoubleClickOrTap("pointer"),
  onMouseClick: (event) => interactions[getActiveMode(mode, event, isZoomDisabled)].onMouseClickOrTap(
    event
  ),
  onMouseDoubleClick: () => interactions[getActiveMode(mode)].onDoubleClickOrTap("pointer"),
  onMouseHover: (event) => interactions[getActiveMode(mode, event, isZoomXDisabled, isPanDisabled)].onMouseHover(event)
});
export {
  getInterfaceInteractions
};
