// @dynatrace/strato-components/esm/charts/core/components/interactions-overlay/utils/get-zoom-interactions.js
import { reset } from "./reset.js";
import {
  panDragEndWithMiddleButton,
  panDragMoveWithMiddleButton,
  panDragStartWithMiddleButton
} from "../interactions/pan/dragInteractions.js";
import {
  zoomDragEnd,
  zoomDragMove,
  zoomDragStart
} from "../interactions/zoom/dragInteractions.js";
import {
  zoomKeyboardDown,
  zoomKeyboardEnter,
  zoomKeyboardEscape,
  zoomKeyboardUp
} from "../interactions/zoom/keyboardInteractions.js";
import {
  zoomMouseClickOrTap,
  zoomMouseHover,
  zoomMouseWheel
} from "../interactions/zoom/mouseInteractions.js";
const getZoomInteractions = ({
  cursorMode,
  selectionArea,
  eventCancelled,
  onSelectEnd,
  lastMousePositionRef,
  pinned,
  onHover,
  onTogglePin,
  setCursorMode,
  height,
  onSelectChange,
  onSelectStart,
  onReset,
  width,
  onZoomChange,
  onPanChange,
  panDeltaRef,
  onPanStart,
  onPanEnd
}) => {
  return {
    onMouseHover: (event) => {
      if (selectionArea) {
        return;
      }
      zoomMouseHover({
        event,
        pinned,
        lastMousePositionRef,
        cursorMode,
        setCursorMode,
        onHover
      });
    },
    onMouseClickOrTap: (event) => zoomMouseClickOrTap({
      event,
      width,
      lastMousePositionRef,
      onZoomChange
    }),
    onDragStart: (event) => zoomDragStart({
      event,
      onSelectStart
    }),
    onDragMove: (event) => zoomDragMove({ event, height, width, onSelectChange }),
    onDragEnd: (event) => zoomDragEnd({
      event,
      height,
      width,
      lastMousePositionRef,
      onSelectEnd,
      onZoomChange
    }),
    onDoubleClickOrTap: (triggeredBy) => {
      onReset?.({
        triggeredBy
      });
      reset(lastMousePositionRef);
    },
    onWheel: (event) => zoomMouseWheel({
      event,
      lastMousePositionRef,
      width,
      onZoomChange,
      onPanChange
    }),
    onKeyEscape: () => zoomKeyboardEscape({
      selectionArea,
      eventCancelled,
      onSelectEnd,
      lastMousePositionRef,
      pinned,
      onHover,
      onTogglePin
    }),
    onKeyEnter: () => zoomKeyboardEnter({
      selectionArea,
      lastMousePositionRef,
      width,
      onSelectEnd,
      onZoomChange,
      onHover
    }),
    onKeyRight: () => {
      return;
    },
    onKeyLeft: () => {
      return;
    },
    onKeyUp: (isTemporaryZoomMode) => {
      if (isTemporaryZoomMode) {
        zoomKeyboardUp({ width, onZoomChange });
      }
    },
    onKeyDown: (isTemporaryZoomMode) => {
      if (isTemporaryZoomMode) {
        zoomKeyboardDown({ width, onZoomChange });
      }
    },
    onDragStartWithMiddleButton: () => panDragStartWithMiddleButton({ panDeltaRef, onPanStart }),
    onDragMoveWithMiddleButton: (event) => panDragMoveWithMiddleButton({ event, panDeltaRef, onPanChange }),
    onDragEndWithMiddleButton: (event) => panDragEndWithMiddleButton({
      event,
      panDeltaRef,
      lastMousePositionRef,
      onPanEnd
    })
  };
};
export {
  getZoomInteractions
};
