// @dynatrace/strato-components/esm/charts/core/components/interactions-overlay/utils/get-explore-interations.js
import {
  exploreDragEnd,
  exploreDragMove,
  exploreDragStart
} from "../interactions/explore/dragInteractions.js";
import {
  exploreKeyboardEnter,
  exploreKeyboardEscape
} from "../interactions/explore/keyboardInteractions.js";
import {
  exploreMouseClickOrTap,
  exploreMouseHover,
  exploreMouseWheel
} from "../interactions/explore/mouseInteractions.js";
import {
  panDragEndWithMiddleButton,
  panDragMoveWithMiddleButton,
  panDragStartWithMiddleButton
} from "../interactions/pan/dragInteractions.js";
import {
  zoomKeyboardDown,
  zoomKeyboardUp
} from "../interactions/zoom/keyboardInteractions.js";
import { reset } from "./reset.js";
const getExploreInteractions = ({
  selectionArea,
  eventCancelled,
  onSelectEnd,
  lastMousePositionRef,
  pinned,
  onHover,
  onTogglePin,
  onModeChange,
  setCursorMode,
  height,
  onSelectChange,
  onSelectStart,
  onReset,
  width,
  onZoomChange,
  isZoomXDisabled,
  onPanChange,
  panDeltaRef,
  onPanStart,
  onPanEnd
}) => {
  return {
    onMouseHover: (event) => exploreMouseHover({
      event,
      pinned,
      selectionArea,
      lastMousePositionRef,
      setCursorMode,
      onHover
    }),
    onMouseClickOrTap: (event) => exploreMouseClickOrTap({
      event,
      selectionArea,
      onHover,
      onTogglePin,
      pinned,
      onSelectEnd,
      lastMousePositionRef
    }),
    onDragStart: (event) => exploreDragStart({ event, onSelectStart }),
    onDragMove: (event) => exploreDragMove({ event, height, width, onSelectChange }),
    onDragEnd: (event) => exploreDragEnd({
      event,
      height,
      width,
      onSelectEnd,
      lastMousePositionRef
    }),
    onDoubleClickOrTap: (triggeredBy) => {
      const mousePosition = lastMousePositionRef.current;
      onReset?.({
        triggeredBy,
        mousePosition
      });
      reset(lastMousePositionRef);
    },
    onWheel: (event) => {
      exploreMouseWheel({ event, onPanChange });
    },
    onKeyEscape: () => exploreKeyboardEscape({
      selectionArea,
      eventCancelled,
      onSelectEnd,
      lastMousePositionRef,
      pinned,
      onHover,
      onTogglePin,
      onModeChange
    }),
    onKeyEnter: () => exploreKeyboardEnter({
      selectionArea,
      lastMousePositionRef,
      width,
      onSelectEnd,
      onZoomChange,
      onHover,
      isZoomXDisabled
    }),
    onKeyRight: () => {
      return;
    },
    onKeyLeft: () => {
      return;
    },
    onKeyUp: (isTemporaryZoomMode) => {
      if (isTemporaryZoomMode) {
        zoomKeyboardUp({ width, onZoomChange });
      }
    },
    onKeyDown: (isTemporaryZoomMode) => {
      if (isTemporaryZoomMode) {
        zoomKeyboardDown({ width, onZoomChange });
      }
    },
    onDragStartWithMiddleButton: () => panDragStartWithMiddleButton({ panDeltaRef, onPanStart }),
    onDragMoveWithMiddleButton: (event) => panDragMoveWithMiddleButton({ event, panDeltaRef, onPanChange }),
    onDragEndWithMiddleButton: (event) => panDragEndWithMiddleButton({
      event,
      panDeltaRef,
      lastMousePositionRef,
      onPanEnd
    })
  };
};
export {
  getExploreInteractions
};
