// @dynatrace/strato-components/charts/core/components/tooltip/types/ChartTooltip.config.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ChartTooltip_config_exports = {};
__export(ChartTooltip_config_exports, {
  ChartTooltipAtoms: () => ChartTooltipAtoms,
  ChartTooltipSlots: () => ChartTooltipSlots,
  ChartTooltipTemplate: () => ChartTooltipTemplate
});
module.exports = __toCommonJS(ChartTooltip_config_exports);
var import_BubbleShape = require("../../shape/BubbleShape/BubbleShape.js");
var import_EmojiShape = require("../../shape/EmojiShape/EmojiShape.js");
var import_IconShape = require("../../shape/IconShape/IconShape.js");
var import_MultiDataPointShape = require("../../shape/MultiDataPointShape.js");
var import_SingleDataPointShape = require("../../shape/SingleDataPointShape.js");
var import_ChipAtom = require("../../tooltip-atoms/components/ChipAtom.js");
var import_DurationChipAtom = require("../../tooltip-atoms/components/DurationChipAtom/DurationChipAtom.js");
var import_PrimaryAtom = require("../../tooltip-atoms/components/PrimaryAtom.js");
var import_SecondaryAtom = require("../../tooltip-atoms/components/SecondaryAtom.js");
var import_SymbolAtom = require("../../tooltip-atoms/components/SymbolAtom.js");
var import_ActionItem = require("../ChartTooltip/components/ActionItem/ActionItem.js");
var import_DurationChip = require("../ChartTooltip/components/DurationChip.js");
var import_EventItem = require("../ChartTooltip/components/EventItem/EventItem.js");
var import_GeoItem = require("../ChartTooltip/components/GeoItem.js");
var import_KeyValue = require("../ChartTooltip/components/KeyValue.js");
var import_RangeHeader = require("../ChartTooltip/components/RangeHeader/RangeHeader.js");
var import_SeriesItem = require("../ChartTooltip/components/SeriesItem/SeriesItem.js");
var import_TextHeader = require("../ChartTooltip/components/TextHeader.js");
var import_TimeFrameHeader = require("../ChartTooltip/components/TimeFrameHeader/TimeFrameHeader.js");
var import_TimestampHeader = require("../ChartTooltip/components/TimestampHeader/TimestampHeader.js");
var import_TooltipText = require("../section-components/atoms/TooltipText.js");
var import_TooltipBody = require("../section-components/TooltipBody.js");
var import_TooltipFooter = require("../section-components/TooltipFooter/TooltipFooter.js");
var import_TooltipHeader = require("../section-components/TooltipHeader/TooltipHeader.js");
var import_TooltipInfoContent = require("../section-components/TooltipInfoContent/TooltipInfoContent.js");
var import_TooltipItem = require("../section-components/TooltipItem/TooltipItem.js");
var import_TooltipSubheader = require("../section-components/TooltipSubheader.js");
var import_TooltipSymbolContent = require("../section-components/TooltipSymbolContent/TooltipSymbolContent.js");
var import_TooltipValueContent = require("../section-components/TooltipValueContent/TooltipValueContent.js");
const ChartTooltipSlots = {
  // section components
  Header: import_TooltipHeader.TooltipHeader,
  Subheader: import_TooltipSubheader.TooltipSubheader,
  Body: import_TooltipBody.TooltipBody,
  Item: import_TooltipItem.TooltipItem,
  Footer: import_TooltipFooter.TooltipFooter,
  // Item subsections
  Symbol: import_TooltipSymbolContent.TooltipSymbolContent,
  Content: import_TooltipInfoContent.TooltipInfoContent,
  Value: import_TooltipValueContent.TooltipValueContent
};
const ChartTooltipAtoms = {
  // atoms
  Chip: import_ChipAtom.ChipAtom,
  DurationChipAtom: import_DurationChipAtom.DurationChipAtom,
  Primary: import_PrimaryAtom.PrimaryAtom,
  Secondary: import_SecondaryAtom.SecondaryAtom,
  Symbol: import_SymbolAtom.SymbolAtom,
  // shapes
  Bubble: import_BubbleShape.BubbleShape,
  Emoji: import_EmojiShape.EmojiShape,
  Icon: import_IconShape.IconShape,
  MultiDataPoint: import_MultiDataPointShape.MultiDataPointShape,
  SingleDataPoint: import_SingleDataPointShape.SingleDataPointShape
};
const ChartTooltipTemplate = {
  GeoItem: import_GeoItem.GeoItem,
  Text: import_TooltipText.TooltipText,
  SeriesItem: import_SeriesItem.SeriesItem,
  ActionItem: import_ActionItem.ActionItem,
  TimeFrameHeader: import_TimeFrameHeader.TimeFrameHeader,
  TimestampHeader: import_TimestampHeader.TimestampHeader,
  RangeHeader: import_RangeHeader.RangeHeader,
  TextHeader: import_TextHeader.TextHeader,
  EventItem: import_EventItem.EventItem,
  KeyValue: import_KeyValue.KeyValue,
  DurationChip: import_DurationChip.DurationChip
};
