// @dynatrace/strato-components/esm/layouts/page/DetailView.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import { forwardRef, useCallback, useRef } from "react";
import { useIntl } from "react-intl";
import * as styles from "./DetailView.sty.js";
import { useKeepMounted } from "./keepMounted.js";
import { usePageContext, useRegisterPagePanel } from "./page-context.js";
import * as panelStyles from "./panel.sty.js";
import { Panel } from "./Panel.js";
import { useWatchBreakpoint } from "./useWatchBreakpoint.js";
import { useId } from "../../core/hooks/useId.js";
import { useMergeRefs } from "../../core/hooks/useMergeRefs.js";
import { Drawer } from "../../overlays/drawer/Drawer.js";
const DetailView = forwardRef(
  (props, forwardedRef) => {
    const {
      breakpoint = 1280,
      className,
      defaultDismissed,
      dismissed,
      minWidth = 320,
      onDismissChange,
      onResize,
      resizable = true,
      preferredWidth = "25%",
      style,
      children,
      keepMounted = false,
      ...restProps
    } = props;
    const panelId = useId();
    const panelName = "details";
    const panelRef = useRef(null);
    const mergedRefs = useMergeRefs([panelRef, forwardedRef]);
    const intl = useIntl();
    const page = usePageContext();
    const panel = useRegisterPagePanel(panelName, {
      ref: panelRef,
      isControlled: dismissed !== void 0,
      isDismissed: dismissed ?? props.defaultDismissed,
      minWidth,
      breakpoint
    });
    const setIsDismissedWithReason = useCallback(
      (value, reason) => {
        if (value === true) {
          page.dismissPanel(panelName);
        } else {
          page.expandPanel(panelName);
        }
        onDismissChange?.(value, reason);
      },
      [onDismissChange, page, panelName]
    );
    const aboveBreakpoint = useWatchBreakpoint({
      breakpoint,
      callback: setIsDismissedWithReason
    });
    const panelChildren = useKeepMounted({ children, keepMounted });
    if (!panel) {
      return null;
    }
    const label = intl.formatMessage({
      defaultMessage: "Detail View",
      id: "rR+xf1BGWvDgZVww"
    });
    const isDisplayedAsDismissed = !aboveBreakpoint || panel.isDismissed;
    return /* @__PURE__ */ jsxs(
      Panel,
      {
        name: panelName,
        id: panelId,
        className: clsx(
          isDisplayedAsDismissed && panelStyles.dismissedPanel,
          styles.details
        ),
        "aria-label": label,
        children: [
          /* @__PURE__ */ jsx(
            Panel.Content,
            {
              ...restProps,
              containerSize: page.containerSize,
              ref: mergedRefs,
              label,
              onResize: (width) => onResize?.(page.getPanelSizes(), width),
              preferredWidth,
              minWidth,
              handle: resizable === true ? "left" : "hidden",
              isDismissed: isDisplayedAsDismissed,
              className: clsx(styles.content, className),
              style,
              children: panelChildren
            }
          ),
          !aboveBreakpoint ? /* @__PURE__ */ jsx(
            Drawer,
            {
              className: panelStyles.drawer,
              isDismissed: panel.isDismissed,
              onDismiss: setIsDismissedWithReason,
              placement: "right",
              width: "wide",
              children: /* @__PURE__ */ jsx("div", { ...restProps, className: panelStyles.drawerContent, children })
            }
          ) : null
        ]
      }
    );
  }
);
DetailView.displayName = "Page.DetailView";
export {
  DetailView
};
