// @dynatrace/strato-components/esm/layouts/page/page-context.js
import {
  createContext,
  useContext,
  useLayoutEffect
} from "react";
const pageContext = createContext({
  containerSize: 0,
  panels: { sidebar: null, main: null, details: null },
  registerPanel: () => void 0,
  removePanel: () => void 0,
  dismissPanel: () => void 0,
  expandPanel: () => void 0,
  togglePanel: () => void 0,
  getPanelSizes: () => void 0
});
function usePageContext() {
  return useContext(pageContext);
}
function useRegisterPagePanel(name, options) {
  const { registerPanel, removePanel, panels } = usePageContext();
  const panel = panels[name];
  useLayoutEffect(() => {
    registerPanel(name, {
      ref: options.ref,
      isControlled: options.isControlled ?? false,
      isDismissed: options.isDismissed ?? false,
      minWidth: options.minWidth ?? 0,
      breakpoint: options.breakpoint ?? 0
    });
    return () => {
      removePanel(name);
    };
  }, [
    name,
    options.breakpoint,
    options.isControlled,
    options.isDismissed,
    options.minWidth,
    options.ref,
    registerPanel,
    removePanel
  ]);
  return panel;
}
export {
  pageContext,
  usePageContext,
  useRegisterPagePanel
};
