// @dynatrace/strato-components/esm/layouts/page/Page.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import {
  useCallback,
  useMemo,
  useReducer,
  useRef
} from "react";
import { DetailView } from "./DetailView.js";
import { Header } from "./Header.js";
import { Main } from "./Main.js";
import {
  pageContext
} from "./page-context.js";
import * as styles from "./Page.sty.js";
import { GAP_SIZE } from "./panel.sty.js";
import { PanelControlButton } from "./PanelControlButton.js";
import { Sidebar } from "./Sidebar.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useContainerSize } from "../../core/hooks/useContainerSize.js";
import { useSlots } from "../../core/hooks/useSlots.js";
const COMPONENT_NAME = "Page";
function handleRegisterAction(state, payload) {
  const { panel, ref, isControlled, isDismissed, minWidth, breakpoint } = payload;
  if (state[panel]) {
    return state;
  }
  const newState = { ...state };
  if (panel === "main") {
    newState[panel] = {
      ref,
      isControlled: false,
      isDismissed: false,
      minWidth: minWidth ?? 0,
      breakpoint: 0
    };
    return newState;
  }
  newState[panel] = {
    ref,
    isControlled,
    isDismissed,
    minWidth: minWidth ?? 0,
    breakpoint: breakpoint ?? 0
  };
  return newState;
}
function handleRemoveAction(state, payload) {
  const { panel } = payload;
  const newState = { ...state, [panel]: null };
  return newState;
}
function handleDismissAction(state, payload) {
  const { panel } = payload;
  if (panel === "main") {
    return state;
  }
  const newState = { ...state };
  if (!newState[panel]) {
    return state;
  }
  const isControlled = newState[panel].isControlled;
  if (isControlled) {
    return state;
  }
  newState[panel] = {
    ...newState[panel],
    isDismissed: true
  };
  return newState;
}
function handleExpandAction(state, payload) {
  const { panel } = payload;
  if (panel === "main") {
    return state;
  }
  const newState = { ...state };
  if (!newState[panel]) {
    return state;
  }
  const isControlled = newState[panel].isControlled;
  if (isControlled) {
    return state;
  }
  newState[panel] = {
    ...newState[panel],
    isDismissed: false
  };
  return newState;
}
function handleToggleAction(state, payload) {
  const { panel } = payload;
  if (panel === "main") {
    return state;
  }
  const newState = { ...state };
  if (!newState[panel]) {
    return state;
  }
  const isControlled = newState[panel].isControlled;
  if (isControlled) {
    return state;
  }
  newState[panel] = {
    ...newState[panel],
    isDismissed: !newState[panel].isDismissed
  };
  return newState;
}
function pageLayoutReducer(state, action) {
  if (action.type === "register") {
    return handleRegisterAction(state, action.payload);
  }
  if (action.type === "remove") {
    return handleRemoveAction(state, action.payload);
  }
  if (action.type === "dismiss") {
    return handleDismissAction(state, action.payload);
  }
  if (action.type === "expand") {
    return handleExpandAction(state, action.payload);
  }
  if (action.type === "toggle") {
    return handleToggleAction(state, action.payload);
  }
  throw Error("invalid action.type");
}
function getGridColumnStyle(panel, max = "auto") {
  const hasDrawer = !window.matchMedia(`(min-width: ${panel.breakpoint}px)`).matches;
  if (hasDrawer || panel.minWidth === void 0 || panel.isDismissed) {
    return `minmax(min-content, ${max})`;
  }
  return `minmax(${panel.minWidth + GAP_SIZE}px, ${max})`;
}
function getGridTemplateStyles(panels) {
  let columns = "";
  let contentAreas = "";
  let headerAreas = "";
  if (panels.main) {
    columns = `${getGridColumnStyle(panels.main, "1fr")} ${columns}`;
    contentAreas = "main";
    headerAreas = "header";
  }
  if (panels.sidebar) {
    columns = `${getGridColumnStyle(panels.sidebar)} ${columns}`;
    contentAreas = `sidebar ${contentAreas}`;
    headerAreas = `header ${headerAreas}`;
  }
  if (panels.details) {
    columns = `${columns} ${getGridColumnStyle(panels.details)}`;
    contentAreas = `${contentAreas} details`;
    headerAreas = `${headerAreas} header`;
  }
  return {
    gridTemplateColumns: columns,
    gridTemplateAreas: `"${headerAreas}" "${contentAreas}"`,
    gridTemplateRows: "min-content minmax(0, 1fr)"
  };
}
const PageRoot = (props) => {
  const { children, style, className, ...remainingProps } = props;
  const [slots, otherChildren] = useSlots(children, {
    header: Header
  });
  const [panels, dispatch] = useReducer(pageLayoutReducer, {
    sidebar: null,
    main: null,
    details: null
  });
  const containerRef = useRef(null);
  const containerSize = useContainerSize(containerRef);
  const getPanelSizes = useCallback(() => {
    const sizes = {};
    for (const [key, panel] of Object.entries(panels)) {
      if (!panel?.ref.current) {
        continue;
      }
      const panelSize = parseFloat(getComputedStyle(panel.ref.current).width);
      sizes[key] = {
        px: panelSize,
        "%": panelSize / containerSize.width * 100
      };
    }
    return sizes;
  }, [containerSize.width, panels]);
  const registerPanel = useCallback(
    (panelName, { ref, isControlled, isDismissed, minWidth, breakpoint }) => {
      dispatch({
        type: "register",
        payload: {
          panel: panelName,
          ref,
          isControlled,
          isDismissed,
          minWidth,
          breakpoint
        }
      });
    },
    []
  );
  const removePanel = useCallback((panelName) => {
    dispatch({ type: "remove", payload: { panel: panelName } });
  }, []);
  const dismissPanel = useCallback((panelName) => {
    dispatch({
      type: "dismiss",
      payload: { panel: panelName }
    });
  }, []);
  const expandPanel = useCallback((panelName) => {
    dispatch({
      type: "expand",
      payload: { panel: panelName }
    });
  }, []);
  const togglePanel = useCallback((panelName) => {
    dispatch({
      type: "toggle",
      payload: { panel: panelName }
    });
  }, []);
  const gridTemplateStyles = getGridTemplateStyles(panels);
  const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
  const ctx = useMemo(
    () => ({
      containerSize: containerSize.width,
      panels,
      registerPanel,
      removePanel,
      dismissPanel,
      expandPanel,
      togglePanel,
      getPanelSizes
    }),
    [
      containerSize.width,
      panels,
      registerPanel,
      removePanel,
      dismissPanel,
      expandPanel,
      togglePanel,
      getPanelSizes
    ]
  );
  return /* @__PURE__ */ jsxs(
    "div",
    {
      ...restProps,
      ...behavioralTrackingProps,
      ref: containerRef,
      className: clsx(styles.page, className),
      style: { ...style, ...gridTemplateStyles },
      "data-dt-component": COMPONENT_NAME,
      children: [
        slots.header,
        /* @__PURE__ */ jsx(pageContext.Provider, { value: ctx, children: otherChildren })
      ]
    }
  );
};
const Page = Object.assign(PageRoot, {
  /** The `Header` component is used inside the `Page` component for displaying content at the very top of the page. */
  Header,
  /** The `Sidebar` component is used inside the `Page` component for displaying content in the corresponding panel. */
  Sidebar,
  /**
   * The `Main` component is used inside the `Page` component for displaying content in the corresponding panel.
   * It is the only compound that is mandatory for rendering the `Page`.
   */
  Main,
  /** The `DetailView` component is used inside the `Page` component for displaying content in the corresponding panel. */
  DetailView,
  /**
   * Use the `PanelControlButton` to get the correct icon for toggling the sidebar / detail view.
   * Depending on where it is rendered and the current state of the panel, the corresponding icon is rendered.
   * You can customize the rendered content by passing any children.
   */
  PanelControlButton
});
export {
  Page
};
