// @dynatrace/strato-components/esm/layouts/page/PanelControlButton.js
import { jsx } from "react/jsx-runtime";
import { useIntl } from "react-intl";
import {
  XmarkIcon,
  OpenSidebarIcon,
  CloseSidebarIcon
} from "@dynatrace/strato-icons";
import { usePageContext } from "./page-context.js";
import { usePanelContext } from "./panel-context.js";
import { Button } from "../../buttons/button/Button.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { isDev } from "../../core/utils/is-node-environment.js";
const PanelControlButton = (props) => {
  const {
    action: actionProp,
    target: targetProp,
    children: childrenProp,
    onClick,
    className: consumerClassName,
    style: consumerStyle,
    ...remainingProps
  } = props;
  const intl = useIntl();
  const pageCtx = usePageContext();
  const panelCtx = usePanelContext();
  const targetPanel = targetProp ?? panelCtx?.name ?? "details";
  const action = actionProp ?? (panelCtx !== null ? "close" : "toggle");
  const panel = pageCtx.panels[targetPanel];
  const isPanelControlled = panel?.isControlled;
  const isPanelDismissed = panel?.isDismissed;
  const shouldOpen = action === "open" || action === "toggle" && isPanelDismissed === true;
  const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
  const defaultDataDtProperties = `state:${shouldOpen ? "expand" : "collapse"}`;
  const {
    "data-dt-name": dataDtName = "ExpandCollapse",
    "data-dt-properties": dataDtProperties = defaultDataDtProperties,
    ...remainingBehavioralTrackingProps
  } = behavioralTrackingProps;
  const handleClick = (event) => {
    if (isPanelControlled && !onClick && isDev()) {
      console.warn(
        "You are using the panel control button with a controlled panel, but didn't provide an 'onClick' callback. Implement the 'onClick' to update the panel state with the panel control button."
      );
      return;
    }
    onClick?.(event);
    if (isPanelControlled) {
      return;
    }
    switch (action) {
      case "open":
        pageCtx.expandPanel(targetPanel);
        break;
      case "close":
        pageCtx.dismissPanel(targetPanel);
        break;
      case "toggle":
        pageCtx.togglePanel(targetPanel);
        break;
    }
  };
  let localizedPanel = intl.formatMessage({
    defaultMessage: "Sidebar",
    id: "MYBFw8pxHVA1QTDj",
    description: "Panel name translation"
  });
  if (targetPanel === "details") {
    localizedPanel = intl.formatMessage({
      defaultMessage: "Detail View",
      id: "0ntbampPIxNnacS8",
      description: "Panel name translation"
    });
  }
  let ariaLabel = intl.formatMessage(
    {
      defaultMessage: "Close {panelName}",
      id: "QsfELKqpB23/KKWM",
      description: "Label for the panel control button."
    },
    { panelName: localizedPanel }
  );
  if (shouldOpen) {
    ariaLabel = intl.formatMessage(
      {
        defaultMessage: "Open {panelName}",
        id: "01469lp26TkXjPEz",
        description: "Label for the panel control button."
      },
      { panelName: localizedPanel }
    );
  }
  let defaultIcon;
  if (shouldOpen) {
    defaultIcon = /* @__PURE__ */ jsx(OpenSidebarIcon, {});
  } else if (action === "toggle") {
    defaultIcon = /* @__PURE__ */ jsx(CloseSidebarIcon, {});
  } else {
    defaultIcon = /* @__PURE__ */ jsx(XmarkIcon, {});
  }
  return /* @__PURE__ */ jsx(
    Button,
    {
      "aria-label": ariaLabel,
      className: consumerClassName,
      style: consumerStyle,
      onClick: handleClick,
      "data-dt-name": dataDtName,
      "data-dt-properties": dataDtProperties,
      ...remainingBehavioralTrackingProps,
      ...restProps,
      children: childrenProp ?? /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(
        "div",
        {
          style: {
            rotate: targetPanel === "sidebar" ? "0deg" : "180deg",
            lineHeight: 0
          },
          children: defaultIcon
        }
      ) })
    }
  );
};
PanelControlButton.displayName = "Page.PanelControlButton";
export {
  PanelControlButton
};
