// @dynatrace/strato-components/esm/layouts/page/Panel.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { useGesture } from "@use-gesture/react";
import clsx from "clsx";
import {
  forwardRef,
  useCallback,
  useEffect,
  useMemo,
  useRef
} from "react";
import { useIntl } from "react-intl";
import { KeepMountedProvider } from "./keepMounted.js";
import { panelContext } from "./panel-context.js";
import * as styles from "./panel.sty.js";
import { useMergeRefs } from "../../core/hooks/useMergeRefs.js";
import { useFocusRing } from "../../core/styles/useFocusRing.js";
const PanelRoot = (props) => {
  const { className, id, style, children, name } = props;
  const ctx = useMemo(() => ({ name }), [name]);
  return /* @__PURE__ */ jsx(panelContext.Provider, { value: ctx, children: /* @__PURE__ */ jsx(KeepMountedProvider, { children: /* @__PURE__ */ jsx(
    "aside",
    {
      ...props,
      id,
      className: clsx(styles.root, className),
      style,
      children
    }
  ) }) });
};
const PanelContent = forwardRef(
  (props, forwardedRef) => {
    const {
      children,
      className,
      handle = "right",
      isDismissed,
      label,
      minWidth,
      onResize,
      preferredWidth,
      style,
      containerSize,
      ...restProps
    } = props;
    const panelRef = useRef(null);
    const panelRefs = useMergeRefs([panelRef, forwardedRef]);
    const handleRef = useRef(null);
    const draggedSize = useRef();
    const wasResized = useRef(false);
    const setPanelSize = useCallback(
      (size) => {
        if (!panelRef.current) {
          return;
        }
        let sizeToApply = Math.floor(size * 1e3) / 1e3;
        draggedSize.current = sizeToApply;
        if (size < 0) {
          return;
        }
        if (minWidth && sizeToApply < minWidth) {
          sizeToApply = minWidth;
        }
        panelRef.current.style.setProperty("width", `${sizeToApply}px`);
      },
      [minWidth]
    );
    const ensurePanelSizeMatchesComputedWidth = useCallback(
      (element, size) => {
        const actualWidth = parseFloat(getComputedStyle(element).width);
        if (size !== actualWidth) {
          setPanelSize(actualWidth);
        }
        return actualWidth;
      },
      [setPanelSize]
    );
    const handleDrag = useCallback(
      (deltaX) => {
        if (!draggedSize.current) {
          return;
        }
        setPanelSize(
          draggedSize.current + deltaX * (handle === "right" ? 1 : -1)
        );
        wasResized.current = true;
      },
      [handle, setPanelSize]
    );
    const handleDragEnd = useCallback(() => {
      if (!panelRef.current || !draggedSize.current) {
        return;
      }
      const size = ensurePanelSizeMatchesComputedWidth(
        panelRef.current,
        draggedSize.current
      );
      onResize?.(size);
    }, [ensurePanelSizeMatchesComputedWidth, onResize]);
    const handleDragStart = useCallback(() => {
      if (!panelRef.current || !draggedSize.current) {
        return;
      }
      ensurePanelSizeMatchesComputedWidth(
        panelRef.current,
        draggedSize.current
      );
    }, [ensurePanelSizeMatchesComputedWidth]);
    useGesture(
      {
        onDragStart: handleDragStart,
        onDragEnd: handleDragEnd,
        onDrag: ({ delta: [deltaX] }) => handleDrag(deltaX)
      },
      {
        target: handleRef,
        eventOptions: { passive: false },
        drag: { filterTaps: true }
      }
    );
    useEffect(
      function initializePanelSize() {
        if (!panelRef.current || wasResized.current || isDismissed) {
          return;
        }
        if (typeof preferredWidth === "number") {
          setPanelSize(preferredWidth);
          return;
        }
        if (preferredWidth?.includes("%")) {
          const size2 = containerSize * (parseFloat(preferredWidth) / 100);
          setPanelSize(size2);
          return;
        }
        if (preferredWidth?.includes("px")) {
          const size2 = parseFloat(preferredWidth);
          setPanelSize(size2);
          return;
        }
        if (typeof preferredWidth === "string") {
          panelRef.current.style.setProperty("width", preferredWidth);
          const size2 = parseFloat(getComputedStyle(panelRef.current).width);
          setPanelSize(size2);
          return;
        }
        const size = parseFloat(getComputedStyle(panelRef.current).width);
        setPanelSize(size);
      },
      [containerSize, preferredWidth, setPanelSize, isDismissed]
    );
    if (isDismissed) {
      return null;
    }
    return /* @__PURE__ */ jsxs(Fragment, { children: [
      /* @__PURE__ */ jsx(
        "div",
        {
          ...restProps,
          ref: panelRefs,
          className: clsx(styles.panelContent, className),
          style: { ...style, minWidth, width: draggedSize.current },
          children
        }
      ),
      /* @__PURE__ */ jsx("div", { className: styles.resizeHandleContainer, children: handle !== "hidden" ? /* @__PURE__ */ jsx(
        ResizeHandle,
        {
          panelName: label,
          panelSize: draggedSize.current,
          ref: handleRef
        }
      ) : null })
    ] });
  }
);
const ResizeHandle = forwardRef(
  (props, ref) => {
    const { panelName, panelSize = 0 } = props;
    const intl = useIntl();
    const { focusProps, focusClassName } = useFocusRing({
      focusWithin: true,
      ignoreModality: true
    });
    return /* @__PURE__ */ jsx(
      "hr",
      {
        ...focusProps,
        ref,
        className: clsx(styles.resizeHandle, focusClassName),
        "aria-valuenow": panelSize,
        "aria-valuetext": `${Math.round(panelSize)}px`,
        "aria-orientation": "vertical",
        "aria-label": intl.formatMessage(
          {
            defaultMessage: "Resize {label} panel",
            id: "Ym3RUHtaLVFsGlaM",
            description: "Label for the resize handle of the panel to resize it with drag and drop or keyboard."
          },
          { label: panelName }
        ),
        tabIndex: 0,
        "data-dt-name": "Resize"
      }
    );
  }
);
const Panel = Object.assign(PanelRoot, {
  Content: PanelContent
});
export {
  Panel
};
