// @dynatrace/strato-components/esm/layouts/page/Sidebar.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import {
  forwardRef,
  useCallback,
  useRef,
  useState
} from "react";
import { useIntl } from "react-intl";
import {
  ChevronSmallLeftDoubleIcon,
  ChevronSmallRightDoubleIcon,
  XmarkIcon
} from "@dynatrace/strato-icons";
import { useKeepMounted } from "./keepMounted.js";
import { usePageContext, useRegisterPagePanel } from "./page-context.js";
import * as panelStyles from "./panel.sty.js";
import { Panel } from "./Panel.js";
import * as styles from "./Sidebar.sty.js";
import { useWatchBreakpoint } from "./useWatchBreakpoint.js";
import { Button } from "../../buttons/button/Button.js";
import { useId } from "../../core/hooks/useId.js";
import { useMergeRefs } from "../../core/hooks/useMergeRefs.js";
import { Drawer } from "../../overlays/drawer/Drawer.js";
import { Tooltip } from "../../overlays/tooltip/Tooltip.js";
const Sidebar = forwardRef(
  (props, forwardedRef) => {
    const {
      breakpoint = 640,
      className,
      defaultDismissed,
      dismissed,
      minWidth = 200,
      onDismissChange,
      onResize,
      preferredWidth = "15%",
      style,
      resizable = true,
      integratedControls,
      children,
      keepMounted = false,
      ...restProps
    } = props;
    const panelId = useId();
    const panelName = "sidebar";
    const panelRef = useRef(null);
    const mergedRefs = useMergeRefs([panelRef, forwardedRef]);
    const intl = useIntl();
    const page = usePageContext();
    const panel = useRegisterPagePanel(panelName, {
      ref: panelRef,
      isControlled: dismissed !== void 0,
      isDismissed: dismissed ?? props.defaultDismissed,
      minWidth,
      breakpoint
    });
    const setIsDismissedWithReason = useCallback(
      (value, reason) => {
        if (value === true) {
          page.dismissPanel(panelName);
        } else {
          page.expandPanel(panelName);
        }
        onDismissChange?.(value, reason);
      },
      [onDismissChange, page, panelName]
    );
    const aboveBreakpoint = useWatchBreakpoint({
      breakpoint,
      callback: setIsDismissedWithReason
    });
    const panelChildren = useKeepMounted({ children, keepMounted });
    if (!panel) {
      return null;
    }
    const label = intl.formatMessage({
      defaultMessage: "Sidebar",
      id: "9RcckxxccNTcNVf9"
    });
    const isDisplayedAsDismissed = !aboveBreakpoint || panel.isDismissed;
    return /* @__PURE__ */ jsxs(
      Panel,
      {
        name: panelName,
        id: panelId,
        className: clsx(
          isDisplayedAsDismissed && panelStyles.dismissedPanel,
          integratedControls && styles.integratedControlsPanel,
          styles.sidebar
        ),
        "aria-label": label,
        children: [
          isDisplayedAsDismissed ? /* @__PURE__ */ jsx("div", { className: panelStyles.gap }) : null,
          /* @__PURE__ */ jsx(
            Panel.Content,
            {
              ...restProps,
              containerSize: page.containerSize,
              ref: mergedRefs,
              label,
              onResize: (width) => onResize?.(page.getPanelSizes(), width),
              preferredWidth,
              minWidth,
              handle: resizable === true ? "right" : "hidden",
              isDismissed: isDisplayedAsDismissed,
              className,
              style,
              children: panelChildren
            }
          ),
          integratedControls ? /* @__PURE__ */ jsx(
            IntegratedControls,
            {
              dismissed: isDisplayedAsDismissed,
              onClick: ({ value, reason }) => {
                setIsDismissedWithReason(value, reason);
              },
              panelId,
              label
            }
          ) : null,
          !aboveBreakpoint ? /* @__PURE__ */ jsxs(
            Drawer,
            {
              className: panelStyles.drawer,
              isDismissed: panel.isDismissed,
              onDismiss: setIsDismissedWithReason,
              placement: "left",
              width: "wide",
              children: [
                integratedControls ? /* @__PURE__ */ jsx("div", { className: styles.integratedControlsDrawerHeader, children: /* @__PURE__ */ jsx(
                  Button,
                  {
                    style: { marginLeft: "auto" },
                    onClick: () => setIsDismissedWithReason(true, "drawer button"),
                    children: /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(XmarkIcon, {}) })
                  }
                ) }) : null,
                /* @__PURE__ */ jsx("div", { ...restProps, className: panelStyles.drawerContent, children })
              ]
            }
          ) : null
        ]
      }
    );
  }
);
Sidebar.displayName = "Page.Sidebar";
const IntegratedControls = ({
  dismissed,
  onClick,
  panelId,
  label
}) => {
  const intl = useIntl();
  const buttonRef = useRef(null);
  const [isTooltipOpen, setIsTooltipOpen] = useState(false);
  const labelText = intl.formatMessage(
    dismissed ? { defaultMessage: "Expand {label} panel", id: "TWLhm3I5du5FpaM+" } : { defaultMessage: "Dismiss {label} panel", id: "SWpMc7FyaE8Uxhzn" },
    { label }
  );
  const buttonProps = dismissed ? {
    "aria-expanded": "false",
    className: styles.expandButton,
    onClick: () => {
      onClick({ value: false, reason: "expand button" });
      setIsTooltipOpen(false);
    },
    variant: "emphasized"
  } : {
    "aria-expanded": "true",
    className: styles.dismissButton,
    onClick: () => {
      onClick({ value: true, reason: "dismiss button" });
      setIsTooltipOpen(false);
    },
    variant: "default"
  };
  return /* @__PURE__ */ jsxs("div", { className: styles.integratedControls, children: [
    dismissed ? /* @__PURE__ */ jsx(
      "button",
      {
        className: styles.hint,
        type: "button",
        onClick: () => {
          onClick({ value: false, reason: "panel hint" });
          setIsTooltipOpen(false);
          buttonRef.current?.focus();
        },
        "aria-controls": panelId,
        "aria-expanded": dismissed ? "false" : "true",
        "aria-label": labelText
      }
    ) : null,
    /* @__PURE__ */ jsx("div", { className: styles.buttonContainer, children: /* @__PURE__ */ jsx(
      Tooltip,
      {
        text: labelText,
        open: isTooltipOpen,
        onOpenChange: setIsTooltipOpen,
        children: /* @__PURE__ */ jsx(
          Button,
          {
            ...buttonProps,
            ref: buttonRef,
            "aria-controls": panelId,
            "aria-label": labelText,
            size: "condensed",
            children: /* @__PURE__ */ jsx(Button.Prefix, { children: dismissed ? /* @__PURE__ */ jsx(ChevronSmallRightDoubleIcon, {}) : /* @__PURE__ */ jsx(ChevronSmallLeftDoubleIcon, {}) })
          }
        )
      }
    ) })
  ] });
};
export {
  IntegratedControls,
  Sidebar
};
