// @dynatrace/strato-components/esm/charts/core/components/legend/categorical/components/ChartLegendItem/ChartLegendItem.sty.js
import "../../../../../../../core/styles/focusRingStatic.css";
import "./ChartLegendItem.css";
import { createRuntimeFn as _7a468 } from "@vanilla-extract/recipes/createRuntimeFn";
var ItemFontStyles = { fontWeight: "var(--dt-typography-text-small-emphasized-weight, 500)", fontSize: "var(--dt-typography-text-small-emphasized-size, 12px)", fontFamily: "var(--dt-typography-text-small-default-family, DynatraceFlow,Roboto,Helvetica,sans-serif)", lineHeight: "var(--dt-typography-text-small-emphasized-line-height, 1.333333333)" };
var baseLegendButton = "_mt8v6g0-3-6-0 _z0fq9b4-3-6-0 _z0fq9bd-3-6-0 _z0fq9b6-3-6-0 _z0fq9b9-3-6-0";
var legendItem = "_mt8v6g9-3-6-0";
var legendItemButton = _7a468({ defaultClassName: "_mt8v6g1-3-6-0", variantClassNames: { isInteractive: { true: "_mt8v6g2-3-6-0", false: "_mt8v6g3-3-6-0" } }, defaultVariants: {}, compoundVariants: [] });
var legendNameButton = _7a468({ defaultClassName: "_mt8v6g4-3-6-0", variantClassNames: { isInteractive: { true: "_mt8v6g5-3-6-0", false: "_mt8v6g6-3-6-0" }, isSubdued: { true: "_mt8v6g7-3-6-0", false: "_mt8v6g8-3-6-0" } }, defaultVariants: {}, compoundVariants: [] });
var legendNameButtonOverride = "_mt8v6gb-3-6-0";
var legendNameWrapper = "_mt8v6ga-3-6-0";
export {
  ItemFontStyles,
  baseLegendButton,
  legendItem,
  legendItemButton,
  legendNameButton,
  legendNameButtonOverride,
  legendNameWrapper
};
