// @dynatrace/strato-components/esm/charts/core/components/legend/categorical/components/ChartLegendItem/ChartLegendItem.js
import { jsx, jsxs } from "react/jsx-runtime";
import { useMemo } from "react";
import { legendItem, legendNameWrapper } from "./ChartLegendItem.sty.js";
import { Tooltip } from "../../../../../../../overlays/tooltip/Tooltip.js";
import { useLegend } from "../../hooks/useLegend.js";
import { useLegendItemActions } from "../../hooks/useLegendItemActions.js";
import { useLegendItemInteractions } from "../../hooks/useLegendItemInteractions.js";
import { useLegendItemState } from "../../hooks/useLegendItemState.js";
import { LEGEND_MESSAGES } from "../../messages.js";
import { ChartLegendIcon } from "../ChartLegendIcon/ChartLegendIcon.js";
import { ChartLegendItemLabel } from "../ChartLegendItemLabel/ChartLegendItemLabel.js";
import { ChartLegendSeriesActions } from "../ChartLegendSeriesActions/ChartLegendSeriesActions.js";
const ChartLegendItem = (props) => {
  const { item, numberOfItems } = props;
  const {
    hiddenItems,
    onLegendIconClick,
    onItemHover,
    onLegendNameClick,
    interactionsDisabled,
    onSeriesAction: actions
  } = useLegend();
  const {
    refIconButton,
    refNameButton,
    open,
    updateOpenState,
    updateHoveredState,
    updateVisible,
    visible,
    isLastItem,
    shouldShowActionsButton,
    isIconHovered
  } = useLegendItemState(
    item,
    interactionsDisabled,
    hiddenItems,
    numberOfItems
  );
  const {
    handleLabelMouseEnter,
    handleIconMouseEnter,
    handleMouseLeave,
    onNameButtonClick,
    onNameButtonKeyDown,
    onIconButtonClick
  } = useLegendItemInteractions(
    item,
    {
      onItemHover,
      onLegendIconClick,
      onLegendNameClick,
      updateHoveredState,
      updateVisible,
      visible
    },
    interactionsDisabled
  );
  const onLegendItemAction = () => {
    onItemHover(void 0);
    updateHoveredState(false);
  };
  const { intl, defaultActions } = useLegendItemActions(
    item,
    onLegendItemAction
  );
  const iconButtonClickHandler = !isLastItem ? onIconButtonClick : void 0;
  const disabled = interactionsDisabled || isLastItem;
  const memoizedActions = useMemo(() => {
    if (open && actions) {
      return actions(item);
    }
  }, [open]);
  return /* @__PURE__ */ jsxs(
    "div",
    {
      onMouseLeave: handleMouseLeave,
      "data-testid": "chart-legend-item",
      className: legendItem,
      children: [
        /* @__PURE__ */ jsx(
          Tooltip,
          {
            "data-testid": disabled ? "chart-legend-item-tooltip-disabled" : "chart-legend-item-tooltip",
            disabled,
            text: visible ? intl.formatMessage(LEGEND_MESSAGES.legendHideSeries) : intl.formatMessage(LEGEND_MESSAGES.legendShowSeries),
            placement: "left",
            children: /* @__PURE__ */ jsx(
              ChartLegendIcon,
              {
                visible,
                hover: isIconHovered,
                disabled: isLastItem,
                color: item.color.Default,
                icon: item.icon,
                buttonRef: refIconButton,
                onMouseEnter: handleIconMouseEnter,
                onButtonClick: iconButtonClickHandler
              }
            )
          }
        ),
        /* @__PURE__ */ jsxs(
          "div",
          {
            onFocus: handleLabelMouseEnter,
            "data-testid": "chart-legend-name-wrapper",
            className: legendNameWrapper,
            children: [
              /* @__PURE__ */ jsx(
                ChartLegendItemLabel,
                {
                  item,
                  visible,
                  buttonRef: refNameButton,
                  onButtonClick: onNameButtonClick,
                  onButtonKeydown: onNameButtonKeyDown,
                  onMouseEnter: handleLabelMouseEnter,
                  onMouseLeave: handleMouseLeave
                }
              ),
              actions && /* @__PURE__ */ jsx(
                ChartLegendSeriesActions,
                {
                  disabled: interactionsDisabled,
                  visible: shouldShowActionsButton,
                  onOpenChange: updateOpenState,
                  isOpen: open,
                  onFocus: handleLabelMouseEnter,
                  onMouseEnter: handleLabelMouseEnter,
                  onMouseLeave: handleMouseLeave,
                  defaultActions,
                  actions: memoizedActions,
                  onBlur: handleMouseLeave
                }
              )
            ]
          }
        )
      ]
    }
  );
};
ChartLegendItem["displayName"] = "ChartLegendItem";
export {
  ChartLegendItem
};
