// @dynatrace/strato-components/charts/core/components/chart-layout/hooks/useResizableLayout.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useResizableLayout_exports = {};
__export(useResizableLayout_exports, {
  useResizableLayout: () => useResizableLayout
});
module.exports = __toCommonJS(useResizableLayout_exports);
var import_react = require("@use-gesture/react");
var import_react2 = require("react");
var import_useChartLayoutRatio = require("./useChartLayoutRatio.js");
var import_get_areas_size_on_drag = require("../utils/get-areas-size-on-drag.js");
var import_get_areas_size = require("../utils/get-areas-size.js");
var import_get_initial_dimensions = require("../utils/get-initial-dimensions.js");
var import_get_resizer_wrapper_size = require("../utils/get-resizer-wrapper-size.js");
var import_orientation = require("../utils/orientation.js");
const initialDimensions = {
  width: 0,
  height: 0
};
const useResizableLayout = (containerWidth, containerHeight, orientation, showResizer = true, showLegend = true, onRatioChange) => {
  const legendResizeContainerRef = (0, import_react2.useRef)(null);
  const legendResizeLineRef = (0, import_react2.useRef)(null);
  const legendContainerRef = (0, import_react2.useRef)(null);
  const { ratio, ratioBoundaries, overrideRatio, legendCandidateSize } = (0, import_useChartLayoutRatio.useChartLayoutRatio)();
  const { minWidth, maxHeight, maxWidth, minHeight } = ratioBoundaries;
  const [areasSizes, setAreasSizes] = (0, import_react2.useState)({
    legend: initialDimensions,
    graph: initialDimensions
  });
  const isVertical = (0, import_orientation.isVertical)(orientation);
  const resizerWrapperSize = showResizer ? (0, import_get_resizer_wrapper_size.getResizerWrapperSize)(isVertical, legendResizeContainerRef) : 0;
  const width = isVertical ? containerWidth : containerWidth - resizerWrapperSize;
  const height = isVertical ? containerHeight - resizerWrapperSize : containerHeight;
  const { height: initialHeight, width: initialWidth } = showLegend ? (0, import_get_initial_dimensions.getInitialDimensions)(
    orientation,
    width,
    height,
    containerWidth,
    containerHeight,
    ratio,
    ratioBoundaries,
    legendCandidateSize,
    overrideRatio
  ) : initialDimensions;
  (0, import_react2.useEffect)(() => {
    const { getAreasHeightSize, getAreasWidthSize } = (0, import_get_areas_size.getAreasSizeHandlers)(
      containerHeight,
      containerWidth,
      resizerWrapperSize
    );
    if (!isVertical && initialWidth && containerWidth) {
      const areasSizes2 = getAreasWidthSize(maxWidth, initialWidth);
      setAreasSizes(areasSizes2);
    } else if (isVertical && initialHeight && containerHeight) {
      const areasSizes2 = getAreasHeightSize(maxHeight, initialHeight);
      setAreasSizes(areasSizes2);
    } else {
      const onlyGraphArea = {
        legend: initialDimensions,
        graph: { width: containerWidth, height: containerHeight }
      };
      setAreasSizes(onlyGraphArea);
    }
  }, [
    containerHeight,
    containerWidth,
    initialHeight,
    initialWidth,
    isVertical,
    maxHeight,
    maxWidth,
    resizerWrapperSize
  ]);
  (0, import_react.useGesture)(
    {
      onDragEnd: () => {
        const legendPercentage = isVertical ? areasSizes.legend.height * 100 / containerHeight : areasSizes.legend.width * 100 / containerWidth;
        onRatioChange?.(Math.round(legendPercentage));
      },
      onDrag: (dragEvent) => {
        const [horizontalDraggedAmount, verticalDraggedAmount] = dragEvent["delta"];
        const currentLegendContainer = legendContainerRef.current;
        if (!currentLegendContainer) {
          return;
        }
        const { getAreasHeightSizeOnDrag, getAreasWidthSizeOnDrag } = (0, import_get_areas_size_on_drag.getAreasSizeOnDragHandlers)(
          containerHeight,
          containerWidth,
          currentLegendContainer,
          resizerWrapperSize
        );
        const newAreasSizes = isVertical ? getAreasHeightSizeOnDrag(
          minHeight,
          maxHeight,
          verticalDraggedAmount
        ) : getAreasWidthSizeOnDrag(
          minWidth,
          maxWidth,
          horizontalDraggedAmount
        );
        setAreasSizes((prevAreasSizes) => {
          if (newAreasSizes) {
            return newAreasSizes;
          }
          const draggedAmount = isVertical ? verticalDraggedAmount : horizontalDraggedAmount;
          const containerSize = isVertical ? containerHeight : containerWidth;
          return (0, import_get_areas_size_on_drag.getDefaultAreasSize)(
            prevAreasSizes,
            draggedAmount,
            containerSize,
            resizerWrapperSize,
            orientation
          );
        });
      }
    },
    {
      target: legendResizeLineRef,
      eventOptions: { passive: false },
      drag: { filterTaps: false },
      enabled: showLegend && showResizer
    }
  );
  return {
    legendWidth: areasSizes.legend.width,
    legendHeight: areasSizes.legend.height,
    graphSize: areasSizes.graph,
    legendResizeLineRef,
    legendContainerRef,
    legendResizeContainerRef
  };
};
