// @dynatrace/strato-components/esm/charts/core/components/tooltip/shared/hooks/useChartTooltip.js
import { useCallback, useEffect, useMemo, useState } from "react";
import { usePopper } from "react-popper";
import { useOverlayStacking } from "../../../../../../core/hooks/useOverlayStacking.js";
import { createDOMRect } from "../../../../../../core/utils/create-dom-rect.js";
import { observeOverlaySize } from "../../../../../../overlays/utils/popper-modifiers.js";
function useChartTooltip(chartTooltipElement, defaultTooltipPlacement = "left") {
  const [virtualElement, setVirtualElement] = useState();
  const tooltipHeight = chartTooltipElement?.getBoundingClientRect().height;
  const popperModifiers = useMemo(() => {
    const modifiers = [
      {
        name: "flip",
        options: {
          fallbackPlacements: [defaultTooltipPlacement]
        }
      },
      {
        name: "preventOverflow",
        options: {
          mainAxis: true
          // true is the default value so maybe it can be removed
        }
      },
      {
        name: "computeStyles",
        options: {
          gpuAcceleration: false
        }
      },
      { name: "eventListeners", enabled: false },
      observeOverlaySize
    ];
    if (tooltipHeight) {
      modifiers.push({
        name: "offset",
        options: {
          offset: () => [
            // 24 Is the SizeLarge token, but since this is needed as a numeric, it is hardcoded here for now.
            Math.round(tooltipHeight / 2 - 24),
            // 16 Is the SizeLarge token, but since this is needed as a numeric, it is hardcoded here for now.
            16
          ]
          // [Aligns the tooltipmarker to the mouse, y-axis margin]
        }
      });
    }
    return modifiers;
  }, [tooltipHeight, defaultTooltipPlacement]);
  const { state, styles } = usePopper(virtualElement, chartTooltipElement, {
    strategy: "fixed",
    placement: defaultTooltipPlacement === "left" ? "right" : "left",
    modifiers: popperModifiers
  });
  const overlayStacking = useOverlayStacking(0, "overlay");
  useEffect(() => {
    const { top, left, right } = styles.popper;
    const display = virtualElement !== void 0 ? "block" : "none";
    chartTooltipElement?.setAttribute(
      "style",
      `display: ${display}; position: fixed; top: ${top}; left: ${left}; right: ${right}; z-index: ${overlayStacking.style.zIndex}`
    );
  }, [
    virtualElement,
    styles.popper,
    chartTooltipElement,
    overlayStacking.style.zIndex
  ]);
  const updateTooltipPosition = useCallback((mousePosition) => {
    if (mousePosition) {
      setVirtualElement({
        getBoundingClientRect: createDOMRect({
          x: mousePosition.x,
          y: mousePosition.y
        })
      });
    } else {
      setVirtualElement(void 0);
    }
  }, []);
  return {
    placement: state?.placement ?? defaultTooltipPlacement,
    updateTooltipPosition
  };
}
export {
  useChartTooltip
};
