// @dynatrace/strato-components/esm/charts/tree-map/components/TreeMapTooltip/TreeMapOverlayTooltip.js
import { jsx, jsxs } from "react/jsx-runtime";
import { useTooltipData } from "../../../core/components/overlay-tooltip/_to-be-removed/utils.js";
import { OverlayTooltip } from "../../../core/components/overlay-tooltip/OverlayTooltip.js";
import { isTooltipDataEmpty } from "../../../core/components/overlay-tooltip/utils.js";
import { useTreeMapCustomTooltip } from "../../hooks/useTreeMapCustomTooltip.js";
import { useTruncationMode } from "../../hooks/useTruncationMode.js";
const TreeMapOverlayTooltip = (props) => {
  const data = useTooltipData().data;
  const truncationMode = useTruncationMode();
  const customTooltip = useTreeMapCustomTooltip();
  if (customTooltip?.hidden || isTooltipDataEmpty(data)) {
    return null;
  }
  const series = data.series[0];
  const meta = data.metadata;
  if (customTooltip?.template) {
    const payload = {
      hierarchyPath: meta.path ?? [series.name],
      value: meta.value,
      childrenCount: meta.children
    };
    return /* @__PURE__ */ jsx(OverlayTooltip, { legacyTemplate: customTooltip.template(payload) });
  }
  return /* @__PURE__ */ jsxs(OverlayTooltip, { truncation: truncationMode, children: [
    /* @__PURE__ */ jsx(OverlayTooltip.Header, { header1: data.header1 }),
    /* @__PURE__ */ jsx(
      OverlayTooltip.SeriesItem,
      {
        name: series.name,
        value: series.value,
        color: series.color,
        actions: props.actionsForItem?.(series)
      }
    )
  ] });
};
TreeMapOverlayTooltip["displayName"] = "TreeMapOverlayTooltip";
export {
  TreeMapOverlayTooltip
};
