// @dynatrace/strato-components/content/accordion/contexts/AccordionContext.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var AccordionContext_exports = {};
__export(AccordionContext_exports, {
  AccordionContextProvider: () => AccordionContextProvider,
  useAccordionContext: () => useAccordionContext
});
module.exports = __toCommonJS(AccordionContext_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
const AccordionContext = (0, import_react.createContext)({
  triggerPosition: "end",
  interactive: true,
  color: "neutral",
  size: "default",
  state: void 0,
  dispatch: void 0
});
function AccordionContextProvider(props) {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(AccordionContext.Provider, { value: props.value, children: props.children });
}
function useAccordionContext() {
  const { triggerPosition, interactive, color, size, state, dispatch } = (0, import_react.useContext)(AccordionContext);
  if (state === void 0 || dispatch === void 0) {
    throw new Error(
      "Accordion context is used outside of an Accordion. Please make sure you are using the correct Accordion components."
    );
  }
  return {
    triggerPosition,
    interactive,
    color,
    size,
    state,
    dispatch
  };
}
