// @dynatrace/strato-components/esm/notifications/notification-settings/NotificationSettings.js
import { jsx, jsxs } from "react/jsx-runtime";
import { forwardRef } from "react";
import { Content } from "./components/Content.js";
import {
  EventNotificationClient,
  isEventNotificationConfig
} from "./components/EventNotificationClient.js";
import { NotificationPermissionInfo } from "./components/PermissionInfo.js";
import {
  isResourceNotificationConfig,
  ResourceNotificationClient
} from "./components/ResourceNotificationClient.js";
import { NotificationToggleButton } from "./components/ToggleButton.js";
import { useAriaLabelingProps } from "../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useSafeForwardProps } from "../../core/hooks/useSafeForwardProps.js";
import { Flex } from "../../layouts/flex/Flex.js";
const COMPONENT_NAME = "NotificationSettings";
function NotificationProvider(props) {
  if (isResourceNotificationConfig(props.type, props.config)) {
    return /* @__PURE__ */ jsx(ResourceNotificationClient, { config: props.config, children: props.children });
  }
  if (isEventNotificationConfig(props.type, props.config)) {
    return /* @__PURE__ */ jsx(EventNotificationClient, { config: props.config, children: props.children });
  }
  return props.children;
}
const NotificationSettings = forwardRef((props, ref) => {
  const {
    config,
    children,
    id,
    type,
    style: consumerStyle,
    className: consumerClassName,
    "data-testid": dataTestId,
    "data-dtrum-mask": dtRumMask,
    "data-dtrum-allow": dtRumAllow,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
  const ariaLabelingProps = useAriaLabelingProps(restProps);
  const forwardProps = useSafeForwardProps(restProps, ariaLabelingProps);
  return /* @__PURE__ */ jsx(NotificationProvider, { type, config, children: /* @__PURE__ */ jsxs(
    Flex,
    {
      id,
      ref,
      flexDirection: "column",
      gap: 12,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dtRumMask,
      "data-dtrum-allow": dtRumAllow,
      ...ariaLabelingProps,
      ...behavioralTrackingProps,
      ...forwardProps,
      "data-dt-component": COMPONENT_NAME,
      children: [
        /* @__PURE__ */ jsx(Content, { children: props.children }),
        /* @__PURE__ */ jsxs(Flex, { alignItems: "center", children: [
          /* @__PURE__ */ jsx(NotificationToggleButton, {}),
          /* @__PURE__ */ jsx(NotificationPermissionInfo, {})
        ] })
      ]
    }
  ) });
});
NotificationSettings.displayName = COMPONENT_NAME;
export {
  NotificationProvider,
  NotificationSettings
};
