// @dynatrace/strato-geo/esm/map/MapView.js
import { jsx, jsxs } from "react/jsx-runtime";
import { forwardRef, useState } from "react";
import {
  _ChartLayout as ChartLayout,
  _coerceSizeValue as coerceSizeValue,
  _useAutoLegendRefresh as useAutoLegendRefresh,
  _OverlayProvider as OverlayProvider,
  _OverlayTooltipProvider as TooltipProvider
} from "@dynatrace/strato-components/charts";
import { _useBehavioralTrackingProps as useBehavioralTrackingProps } from "@dynatrace/strato-components/core";
import { MapLegendRenderer } from "./components/legend/MapLegendRenderer.js";
import { MapContent } from "./components/MapContent.js";
import { MapUnavailable } from "./components/MapUnavailable.js";
import {
  DEFAULT_LEGEND_SIZES,
  DEFAULT_MAP_HEIGHT,
  DEFAULT_TRUNCATION_MODE,
  MAP_VIEW_ARIA_LABEL
} from "./constants.js";
import { FormatterContext } from "./contexts/formatter.context.js";
import { LayerIdsContext } from "./contexts/layer-ids.context.js";
import { MapBaseLayerFeaturesContext } from "./contexts/map-base-layer.context.js";
import { MapConfigurationContext } from "./contexts/map-configuration.context.js";
import { MapLoadingContext } from "./contexts/map-loading.context.js";
import { MapRawDataContext } from "./contexts/map-raw-data.context.js";
import { MapTruncationModeContext } from "./contexts/map-truncation-mode.context.js";
import { useLoadMapBaseLayer } from "./hooks/use-load-map-base-layer.js";
import { ColorScaleProvider } from "./providers/color-scale.provider.js";
import { LayerColorStrategyProvider } from "./providers/layer-color-strategy.provider.js";
import { DownloadCSV } from "./slots/DownloadCSV.js";
import { ErrorStateSlot } from "./slots/states/ErrorStateSlot.js";
import { Toolbar } from "./slots/Toolbar.js";
import { Zoom } from "./slots/Zoom.js";
import { MapStoreProvider } from "./store/map-store.provider.js";
import { isThresholdLegend } from "./utils/build-scale-from-legend-config.js";
import { extractLayersData } from "./utils/extract-layers-data.js";
import { getMapStatesTemplates } from "./utils/get-map-states-template.js";
import { iterateConfigSlots } from "./utils/iterate-config-slots.js";
import { setLayersId } from "./utils/set-layers-id.js";
import "./styles/react-mapgl-styles.css";
const _MapView = forwardRef(
  (props, forwardedRef) => {
    const [behavioralTrackingProps] = useBehavioralTrackingProps(props);
    const {
      children,
      loading = false,
      style: costumerStyle,
      className: customerClassNames,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "data-testid": dataTestId,
      formatter,
      truncationMode,
      ...remaining
    } = props;
    const { layerIds, parsedChildren, valueAccessors } = setLayersId(children);
    const {
      flattenData: layersData,
      categories,
      legendDomain
    } = extractLayersData(parsedChildren, valueAccessors);
    const config = iterateConfigSlots(children, legendDomain);
    const getMaxRange = () => {
      if (config.legend && isThresholdLegend(config.legend)) {
        return config.legend.ranges.at(-1)?.to ?? legendDomain[1];
      }
      return legendDomain[1];
    };
    const dataMax = getMaxRange();
    const isLegendHidden = !config.legend || !!config.legend.hidden;
    const legendPosition = config.legend?.position;
    const legendRatio = config.legend?.ratio;
    const legendOnRatioChange = config.legend?.onRatioChange;
    const { isMapEnabled, isFetchingFeatures, baseFeatureCollection, error } = useLoadMapBaseLayer(config.baseLayer);
    const [isMapLoaded, setIsMapLoaded] = useState(false);
    const { errorState } = getMapStatesTemplates(children);
    const chartHeight = coerceSizeValue(props.height) || DEFAULT_MAP_HEIGHT;
    const containerStyles = { width: "100%", ...costumerStyle };
    const chartLayoutRef = useAutoLegendRefresh(categories.join(":"));
    const mapUnavailable = !isFetchingFeatures && !isMapEnabled || error;
    const isLoading = isFetchingFeatures || !isMapLoaded;
    return /* @__PURE__ */ jsx(
      "div",
      {
        style: containerStyles,
        className: customerClassNames,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        "data-testid": dataTestId,
        role: "img",
        "aria-label": MAP_VIEW_ARIA_LABEL,
        ...behavioralTrackingProps,
        // TODO: Error boundaries global fix needed - this error state only purpose is to make map not rendering when there error in response
        children: mapUnavailable ? /* @__PURE__ */ jsx(
          MapUnavailable,
          {
            isFetchingFeatures,
            isDisabled: !isMapEnabled,
            chartHeight,
            error
          }
        ) : /* @__PURE__ */ jsx(MapLoadingContext.Provider, { value: loading, children: /* @__PURE__ */ jsx(
          MapBaseLayerFeaturesContext.Provider,
          {
            value: baseFeatureCollection,
            children: /* @__PURE__ */ jsx(MapConfigurationContext.Provider, { value: config, children: /* @__PURE__ */ jsx(LayerIdsContext.Provider, { value: layerIds, children: /* @__PURE__ */ jsx(
              MapTruncationModeContext.Provider,
              {
                value: truncationMode ?? DEFAULT_TRUNCATION_MODE,
                children: /* @__PURE__ */ jsx(FormatterContext.Provider, { value: formatter, children: /* @__PURE__ */ jsx(MapRawDataContext.Provider, { value: layersData, children: /* @__PURE__ */ jsx(OverlayProvider, { children: /* @__PURE__ */ jsx(TooltipProvider, { children: /* @__PURE__ */ jsx(MapStoreProvider, { children: /* @__PURE__ */ jsx(
                  ColorScaleProvider,
                  {
                    categories,
                    dataMax,
                    children: /* @__PURE__ */ jsx(LayerColorStrategyProvider, { children: /* @__PURE__ */ jsxs(
                      ChartLayout,
                      {
                        ref: chartLayoutRef,
                        chartHeight,
                        errorState,
                        showLoader: loading || isLoading,
                        children: [
                          /* @__PURE__ */ jsx(ChartLayout.Graph, { children: baseFeatureCollection && /* @__PURE__ */ jsx(
                            MapContent,
                            {
                              ref: forwardedRef,
                              onMapLoad: () => setIsMapLoaded(true),
                              ...remaining,
                              children: parsedChildren
                            }
                          ) }),
                          !isLegendHidden && /* @__PURE__ */ jsx(
                            ChartLayout.Legend,
                            {
                              position: legendPosition,
                              ratio: legendRatio,
                              ratioBoundaries: DEFAULT_LEGEND_SIZES,
                              onResize: legendOnRatioChange,
                              children: /* @__PURE__ */ jsx(MapLegendRenderer, {})
                            }
                          )
                        ]
                      }
                    ) })
                  }
                ) }) }) }) }) })
              }
            ) }) })
          }
        ) })
      }
    );
  }
);
_MapView["displayName"] = "MapView";
const MapView = Object.assign(_MapView, {
  ErrorState: ErrorStateSlot,
  Toolbar,
  DownloadCSV,
  Zoom
});
export {
  MapView
};
