// @dynatrace/strato-components/charts/core/components/tooltip/section-components/atoms/TooltipText.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TooltipText_exports = {};
__export(TooltipText_exports, {
  TooltipText: () => TooltipText
});
module.exports = __toCommonJS(TooltipText_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_OverlayStackingProvider = require("../../../../../../core/providers/OverlayStackingProvider.js");
var import_Tooltip = require("../../../../../../overlays/tooltip/Tooltip.js");
var import_constants = require("../../../../constants.js");
var import_useTruncationDetector = require("../../../../hooks/useTruncationDetector.js");
var import_ChartTextEllipsis = require("../../../text-ellipsis/ChartTextEllipsis.js");
var import_PrimaryAtom = require("../../../tooltip-atoms/components/PrimaryAtom.js");
var import_SecondaryAtom = require("../../../tooltip-atoms/components/SecondaryAtom.js");
const TooltipText = (props) => {
  const {
    children: text,
    variant = "primary",
    truncationMode = import_constants.DEFAULT_TRUNCATION_MODE,
    ...rest
  } = props;
  const textRef = (0, import_react.useRef)(null);
  const isTruncated = (0, import_useTruncationDetector.useTruncationDetector)(textRef, text);
  const intl = (0, import_react_intl.useIntl)();
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayStackingProvider.OverlayStackingProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Tooltip.Tooltip, { text, disabled: !isTruncated, children: variant === "primary" ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_PrimaryAtom.PrimaryAtom, { ...rest, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ChartTextEllipsis.ChartTextEllipsis,
    {
      ref: textRef,
      truncationMode,
      "aria-label": intl.formatMessage({
        defaultMessage: "Primary series name",
        id: "OwC/n+tx4kFWA3l3"
      }),
      children: text
    }
  ) }) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_SecondaryAtom.SecondaryAtom, { ...rest, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ChartTextEllipsis.ChartTextEllipsis,
    {
      ref: textRef,
      truncationMode,
      "aria-label": intl.formatMessage({
        defaultMessage: "Secondary series name",
        id: "QbeiZKf6kqgyTp3m"
      }),
      children: text
    }
  ) }) }) });
};
TooltipText["displayName"] = "TooltipText";
