// @dynatrace/strato-components/esm/charts/core/hooks/useDragElement.js
import { useGesture } from "@use-gesture/react";
import { useEffect, useLayoutEffect, useState, useRef } from "react";
import { fitBetween } from "../utils/fitBetween.js";
function useDragElement(toolbarElement, maxBounds, initialPosition) {
  const [isDragging, setIsDragging] = useState(false);
  const [didDrag, setDidDrag] = useState(false);
  const triggerDragRef = useRef(null);
  const [currentPosition, setCurrentPosition] = useState(initialPosition);
  const [lastIdlePosition, setLastIdlePosition] = useState(initialPosition);
  useLayoutEffect(() => {
    setCurrentPosition((prevCurrentPosition) => {
      const right = fitBetween(prevCurrentPosition.right, 0, maxBounds.right);
      const top = fitBetween(prevCurrentPosition.top, 0, maxBounds.top);
      return { right, top };
    });
    setLastIdlePosition((prevIdlePosition) => {
      const right = fitBetween(prevIdlePosition.right, 0, maxBounds.right);
      const top = fitBetween(prevIdlePosition.top, 0, maxBounds.top);
      return { right, top };
    });
  }, [maxBounds]);
  useEffect(() => {
    if (didDrag) {
      return;
    }
    setCurrentPosition(initialPosition);
    setLastIdlePosition(initialPosition);
  }, [initialPosition.top, initialPosition.right]);
  useEffect(() => {
    const { top, right } = currentPosition;
    if (toolbarElement) {
      toolbarElement.style.right = `${right}px`;
      toolbarElement.style.top = `${top}px`;
    }
  }, [toolbarElement, currentPosition]);
  useGesture(
    {
      onDrag: (event) => {
        const [movedX, movedY] = event.movement;
        if (event.intentional) {
          const x = lastIdlePosition.right - movedX;
          const y = lastIdlePosition.top + movedY;
          const right = fitBetween(x, 0, maxBounds.right);
          const top = fitBetween(y, 0, maxBounds.top);
          setCurrentPosition({ right, top });
          setIsDragging(true);
          setDidDrag(true);
        }
      },
      onDragEnd: () => {
        setLastIdlePosition(currentPosition);
        setIsDragging(false);
      }
    },
    {
      target: triggerDragRef,
      eventOptions: { passive: false },
      drag: { filterTaps: true, threshold: 8 }
    }
  );
  return { isDragging, triggerDragRef, didDrag, setDidDrag };
}
export {
  useDragElement
};
