// @dynatrace/strato-components/esm/charts/core/hooks/useFontSizeForContainer.js
import { useMemo } from "react";
import { FONT_BASE, SUFFIX_RATIO } from "../constants.js";
import { measureFontBaseline } from "../utils/measure-font-baseline.js";
const SPACING_RATIO_THRESHOLD = 1.5;
const MAX_HORIZONTAL_WEIGHT = 0.8;
const BASE_HORIZONTAL_WEIGHT = 0.5;
const WEIGHT_ADJUSTMENT_FACTOR = 0.1;
const VERTICAL_CONSTRAINT_FACTOR = 1.1;
function useFontSizeForContainer({
  container,
  text,
  element,
  suffix = "",
  suffixElement
}) {
  const style = element ? getComputedStyle(element) : {};
  const valueBaseline = useMemo(
    () => measureFontBaseline({
      text,
      fontStyle: style.fontStyle,
      fontWeight: style.fontWeight,
      fontFamily: style.fontFamily
    }),
    [text, style.fontStyle, style.fontWeight, style.fontFamily]
  );
  const suffixStyle = suffixElement ? getComputedStyle(suffixElement) : {};
  const suffixBaseline = useMemo(
    () => measureFontBaseline({
      text: suffix,
      fontStyle: suffixStyle.fontStyle,
      fontWeight: suffixStyle.fontWeight,
      fontFamily: suffixStyle.fontFamily
    }),
    [
      suffix,
      suffixStyle.fontStyle,
      suffixStyle.fontWeight,
      suffixStyle.fontFamily
    ]
  );
  function calculateWeightedFontSize(fontSizeFitHorizontal, fontSizeFitVertical, spacingRatio) {
    const horizontalWeight = Math.min(
      MAX_HORIZONTAL_WEIGHT,
      BASE_HORIZONTAL_WEIGHT + (spacingRatio - SPACING_RATIO_THRESHOLD) * WEIGHT_ADJUSTMENT_FACTOR
    );
    const verticalWeight = 1 - horizontalWeight;
    const weightedAverage = fontSizeFitHorizontal * horizontalWeight + fontSizeFitVertical * verticalWeight;
    return Math.min(
      weightedAverage,
      fontSizeFitVertical * VERTICAL_CONSTRAINT_FACTOR
    );
  }
  function calculateStandardFontSize(fontSizeFitHorizontal, fontSizeFitVertical) {
    return Math.min(fontSizeFitHorizontal, fontSizeFitVertical);
  }
  if (valueBaseline) {
    const suffixWidth = suffixBaseline ? suffixBaseline.width / FONT_BASE * SUFFIX_RATIO : 0;
    const fontSizeFitHorizontal = container.width / (valueBaseline.width / FONT_BASE + suffixWidth);
    const fontSizeFitVertical = container.height / (valueBaseline.height / FONT_BASE);
    const spacingRatio = fontSizeFitHorizontal / fontSizeFitVertical;
    let preferredFontSize;
    if (spacingRatio > SPACING_RATIO_THRESHOLD) {
      preferredFontSize = calculateWeightedFontSize(
        fontSizeFitHorizontal,
        fontSizeFitVertical,
        spacingRatio
      );
    } else {
      preferredFontSize = calculateStandardFontSize(
        fontSizeFitHorizontal,
        fontSizeFitVertical
      );
    }
    return isFinite(preferredFontSize) ? preferredFontSize : 0;
  }
  return 0;
}
export {
  useFontSizeForContainer
};
