// @dynatrace/strato-components/esm/charts/core/hooks/useObserveContainerSize.js
import { toNumber } from "lodash-es";
import { useCallback, useRef, useState } from "react";
function calculateNodeDimensions(node) {
  const getNodePaddings = (node2) => {
    const coercePadding = (padding2) => toNumber(padding2.match(/^\d+/g)?.join("") ?? 0);
    const computedElStyle = getComputedStyle(node2);
    return {
      top: coercePadding(computedElStyle.paddingTop),
      bottom: coercePadding(computedElStyle.paddingBottom),
      left: coercePadding(computedElStyle.paddingLeft),
      right: coercePadding(computedElStyle.paddingRight)
    };
  };
  const { width, height } = node.getBoundingClientRect();
  const padding = getNodePaddings(node);
  const { height: wrapperHeight } = node.parentElement?.getBoundingClientRect() ?? { height: 0 };
  return {
    width: width - padding.left - padding.right,
    height: height - padding.top - padding.bottom,
    padding,
    wrapperHeight: wrapperHeight - padding.top - padding.bottom
  };
}
const useObserveContainerSize = () => {
  const [dimensions, setDimensions] = useState({
    width: 0,
    height: 0,
    padding: { top: 0, bottom: 0, left: 0, right: 0 },
    wrapperHeight: 0
  });
  const observerCleanupRef = useRef(void 0);
  const ref = useCallback((node) => {
    if (node === null) {
      if (observerCleanupRef.current) {
        observerCleanupRef.current();
      }
      return;
    }
    setDimensions(calculateNodeDimensions(node));
    const observer = new ResizeObserver(() => {
      setDimensions(calculateNodeDimensions(node));
    });
    observerCleanupRef.current = () => observer.unobserve(node);
    observer.observe(node);
  }, []);
  return [dimensions, ref];
};
export {
  useObserveContainerSize
};
