// @dynatrace/strato-components/content/avatar/Avatar.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Avatar_exports = {};
__export(Avatar_exports, {
  Avatar: () => Avatar
});
module.exports = __toCommonJS(Avatar_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_Avatar_css = require("./Avatar.sty.js");
var import_Label = require("./Label.js");
var import_Subtitle = require("./Subtitle.js");
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useSafeForwardProps = require("../../core/hooks/useSafeForwardProps.js");
var import_useSlots = require("../../core/hooks/useSlots.js");
var import_focusRingStatic_css = require("../../core/styles/focusRingStatic.sty.js");
var import_sanitize_symbol = require("../../core/utils/sanitize-symbol.js");
var import_Flex = require("../../layouts/flex/Flex.js");
var import_textStyle_css = require("../../styles/textStyle.sty.js");
const COMPONENT_NAME = "Avatar";
const InternalAvatar = /* @__PURE__ */ (0, import_react.forwardRef)(
  (props, ref) => {
    const {
      children,
      size = "default",
      abbreviation,
      disabled = false,
      as,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
    const ariaLabelingProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(restProps);
    const forwardProps = (0, import_useSafeForwardProps.useSafeForwardProps)(restProps, ariaLabelingProps);
    const [slots] = (0, import_useSlots.useSlots)(children, {
      label: import_Label.Label,
      subtitle: import_Subtitle.Subtitle
    });
    const Abbreviation = as ?? "div";
    const remainingConsumerClassname = (consumerClassName ?? "").replace(
      /rovingFocusElement/g,
      ""
    );
    const hasRovingFocus = (consumerClassName ?? "").includes(
      "rovingFocusElement"
    );
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "div",
      {
        className: (0, import_clsx.default)(
          import_Avatar_css.avatarWrapperCSS,
          (0, import_textStyle_css.textStyle)({ textStyle: "base" }),
          remainingConsumerClassname
        ),
        style: consumerStyle,
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        ...behavioralTrackingProps,
        "data-dt-component": COMPONENT_NAME,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { gap: size === "small" ? 4 : 8, alignItems: "center", minWidth: 0, children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            Abbreviation,
            {
              ref,
              className: (0, import_clsx.default)(
                {
                  [import_Avatar_css.avatarDisabledCSS]: disabled,
                  rovingFocusElement: hasRovingFocus
                },
                (0, import_Avatar_css.avatarAbbreviationCSS)({ size, disabled }),
                (0, import_textStyle_css.textStyle)({ textStyle: size === "small" ? "small" : "base" }),
                (0, import_focusRingStatic_css.focusRingStatic)({ disabled, when: "focus-within-visible" })
              ),
              ...ariaLabelingProps,
              ...forwardProps,
              children: (0, import_sanitize_symbol._sanitizeSymbol)(abbreviation, size === "default" ? 2 : void 0)
            }
          ),
          (slots.label || slots.subtitle) && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { flexDirection: "column", gap: 0, minWidth: 0, children: [
            slots.label,
            slots.subtitle
          ] })
        ] })
      }
    );
  }
);
InternalAvatar.displayName = COMPONENT_NAME;
const Avatar = Object.assign(InternalAvatar, {
  Label: import_Label.Label,
  Subtitle: import_Subtitle.Subtitle
});
