// @dynatrace/strato-components/content/avatar/Label.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Label_exports = {};
__export(Label_exports, {
  Label: () => Label
});
module.exports = __toCommonJS(Label_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_can_receive_default_text_format = require("../../core/utils/can-receive-default-text-format.js");
var import_Text = require("../../typography/text/Text.js");
const Label = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    children,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId
  } = props;
  const childIsText = (0, import_can_receive_default_text_format.canReceiveDefaultTextFormat)(children);
  return childIsText ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Text.Text,
    {
      as: "div",
      textStyle: "base",
      ref: forwardedRef,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      children
    }
  ) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      "data-testid": dataTestId,
      ref: forwardedRef,
      className: consumerClassName,
      style: consumerStyle,
      children
    }
  );
});
Label.displayName = "Avatar.Label";
