// @dynatrace/strato-components/charts/core/components/annotations/AnnotationsAtom.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var AnnotationsAtom_exports = {};
__export(AnnotationsAtom_exports, {
  AnnotationsAtom: () => AnnotationsAtom,
  _AnnotationsAtom: () => _AnnotationsAtom
});
module.exports = __toCommonJS(AnnotationsAtom_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_AnnotationsAtom_css = require("./AnnotationsAtom.sty.js");
var import_AnnotationsCanvasLayout = require("./components/canvas/AnnotationsCanvasLayout.js");
var import_CanvasSvgToImages_Provider = require("./components/canvas/providers/CanvasSvgToImages.Provider.js");
var import_AnnotationsOverlays = require("./components/overlays/AnnotationsOverlays.js");
var import_constants = require("./constants.js");
var import_AnnotationsConfig_context = require("./contexts/AnnotationsConfig.context.js");
var import_MarkerCoordinates_context = require("./contexts/MarkerCoordinates.context.js");
var import_MarkerEventHandlers_context = require("./contexts/MarkerEventHandlers.context.js");
var import_MarkersByTrack_context = require("./contexts/MarkersByTrack.context.js");
var import_TrackEventHandlers_context = require("./contexts/TrackEventHandlers.context.js");
var import_MarkerAtom = require("./slots/MarkerAtom.js");
var import_TrackAtom = require("./slots/TrackAtom.js");
var import_aggregate_markers_coords = require("./utils/aggregate-markers-coords.js");
var import_extract_visible_tracks = require("./utils/extract-visible-tracks.js");
var import_get_annotations_tracks_height = require("./utils/get-annotations-tracks-height.js");
const _AnnotationsAtom = (props) => {
  const {
    width,
    children,
    showLabels = true,
    labelsWidth = import_constants.DEFAULT_LABELS_WIDTH
  } = props;
  const trackWidth = width - labelsWidth;
  const tracks = (0, import_extract_visible_tracks.extractVisibleTracks)(children);
  const tracksQuantity = import_react.Children.count(tracks);
  const totalTracksHeight = (0, import_get_annotations_tracks_height.getAnnotationsTracksHeight)(tracksQuantity);
  const {
    markersPerTrack,
    markerEventHandlers,
    trackEventHandlers,
    icons,
    tracksConfig
  } = (0, import_aggregate_markers_coords.aggregateMarkersCoords)(tracks, trackWidth);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      className: import_AnnotationsAtom_css.wrapper,
      style: {
        height: totalTracksHeight
      },
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_AnnotationsConfig_context.AnnotationsConfigContext.Provider,
        {
          value: {
            showLabels,
            labelsWidth,
            trackWidth
          },
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_MarkersByTrack_context.MarkersByTrackContext.Provider, { value: tracksConfig, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_MarkerCoordinates_context.MarkerCoordinatesContext.Provider, { value: markersPerTrack, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_MarkerEventHandlers_context.MarkerEventHandlersContext.Provider, { value: markerEventHandlers, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TrackEventHandlers_context.TrackEventHandlersContext.Provider, { value: trackEventHandlers, children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_CanvasSvgToImages_Provider.CanvasSvgToImagesProvider, { icons, children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_AnnotationsOverlays.AnnotationsOverlays,
              {
                height: totalTracksHeight - import_constants.TRACK_SEPARATOR_STROKE_WIDTH
              }
            ),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_AnnotationsCanvasLayout.AnnotationsCanvasLayout,
              {
                width,
                height: totalTracksHeight
              }
            )
          ] }) }) }) }) })
        }
      )
    }
  );
};
const AnnotationsAtom = Object.assign(_AnnotationsAtom, {
  TrackAtom: import_TrackAtom.TrackAtom,
  MarkerAtom: import_MarkerAtom.MarkerAtom
});
AnnotationsAtom.displayName = "AnnotationsAtom";
