// @dynatrace/strato-components/charts/core/components/annotations/constants.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var constants_exports = {};
__export(constants_exports, {
  DEFAULT_ICON_COLOR: () => DEFAULT_ICON_COLOR,
  DEFAULT_INTEGRATION_TEXT_OVERFLOW: () => DEFAULT_INTEGRATION_TEXT_OVERFLOW,
  DEFAULT_LABELS_OFFSET: () => DEFAULT_LABELS_OFFSET,
  DEFAULT_LABELS_WIDTH: () => DEFAULT_LABELS_WIDTH,
  DEFAULT_MARKER_COLOR: () => DEFAULT_MARKER_COLOR,
  DEFAULT_MARKER_HIDDEN: () => DEFAULT_MARKER_HIDDEN,
  DEFAULT_MARKER_PRIORITY: () => DEFAULT_MARKER_PRIORITY,
  DEFAULT_SHOW_LABELS: () => DEFAULT_SHOW_LABELS,
  DEFAULT_TEXT_OVERFLOW: () => DEFAULT_TEXT_OVERFLOW,
  DEFAULT_TOOLTIP_MAX_LINES: () => DEFAULT_TOOLTIP_MAX_LINES,
  DEFAULT_TOOLTIP_Y_OFFSET: () => DEFAULT_TOOLTIP_Y_OFFSET,
  DEFAULT_TRACK_HIDDEN: () => DEFAULT_TRACK_HIDDEN,
  DEFAULT_TRUNCATION_MODE: () => DEFAULT_TRUNCATION_MODE,
  EMPHASIZED_TRACK_SEPARATOR_COLOR: () => EMPHASIZED_TRACK_SEPARATOR_COLOR,
  FOREIGN_OBJECT_THRESHOLD: () => FOREIGN_OBJECT_THRESHOLD,
  LABEL_OFFSET: () => LABEL_OFFSET,
  MARGIN_TOP_STROKE: () => MARGIN_TOP_STROKE,
  MARKER_BORDER_RADIUS: () => MARKER_BORDER_RADIUS,
  MARKER_CONTENT_TO_BORDER_PADDING: () => MARKER_CONTENT_TO_BORDER_PADDING,
  MARKER_ICON_OFFSET: () => MARKER_ICON_OFFSET,
  MARKER_LABEL_PADDING: () => MARKER_LABEL_PADDING,
  MARKER_OPACITY: () => MARKER_OPACITY,
  MARKER_PROPS: () => MARKER_PROPS,
  MARKER_RADIUS: () => MARKER_RADIUS,
  MARKER_STROKE_COLOR: () => MARKER_STROKE_COLOR,
  MARKER_STROKE_WIDTH: () => MARKER_STROKE_WIDTH,
  MINIMUM_MARKER_WIDTH: () => MINIMUM_MARKER_WIDTH,
  OUTER_CIRCLE_RADIUS: () => OUTER_CIRCLE_RADIUS,
  OUTLINE_STROKE_WIDTH: () => OUTLINE_STROKE_WIDTH,
  PILL_MARKER_HEIGHT: () => PILL_MARKER_HEIGHT,
  SCROLLBAR_THRESHOLD: () => SCROLLBAR_THRESHOLD,
  SCROLLBAR_WIDTH: () => SCROLLBAR_WIDTH,
  TRACK_HEIGHT: () => TRACK_HEIGHT,
  TRACK_PADDING_Y: () => TRACK_PADDING_Y,
  TRACK_SEPARATOR_COLOR: () => TRACK_SEPARATOR_COLOR,
  TRACK_SEPARATOR_STROKE_WIDTH: () => TRACK_SEPARATOR_STROKE_WIDTH,
  TestIds: () => TestIds
});
module.exports = __toCommonJS(constants_exports);
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
const DEFAULT_LABELS_WIDTH = 60;
const DEFAULT_LABELS_OFFSET = 20;
const TRACK_HEIGHT = 24;
const TRACK_SEPARATOR_COLOR = import_colors.default.Border.Neutral.Default;
const EMPHASIZED_TRACK_SEPARATOR_COLOR = import_colors.default.Border.Neutral.Accent;
const TRACK_SEPARATOR_STROKE_WIDTH = 1;
const TRACK_PADDING_Y = 1;
const LABEL_OFFSET = 12;
const PILL_MARKER_HEIGHT = 20;
const OUTLINE_STROKE_WIDTH = 2;
const DEFAULT_MARKER_COLOR = import_colors.default.Charts.Categorical.Color01.Default;
const MARKER_RADIUS = 10;
const MINIMUM_MARKER_WIDTH = 3;
const MARKER_OPACITY = 0.85;
const MARKER_STROKE_WIDTH = 1;
const MARKER_STROKE_COLOR = import_colors.default.Background.Base.Default;
const OUTER_CIRCLE_RADIUS = MARKER_RADIUS + OUTLINE_STROKE_WIDTH;
const SCROLLBAR_THRESHOLD = 15;
const SCROLLBAR_WIDTH = 8;
const DEFAULT_ICON_COLOR = import_colors.default.Text.Neutral.OnAccent.Default;
const MARKER_PROPS = {
  opacity: MARKER_OPACITY,
  stroke: MARKER_STROKE_COLOR,
  strokeWidth: MARKER_STROKE_WIDTH
};
const DEFAULT_TRACK_HIDDEN = false;
const DEFAULT_MARKER_PRIORITY = 0;
const DEFAULT_MARKER_HIDDEN = false;
const DEFAULT_TOOLTIP_Y_OFFSET = 40;
const DEFAULT_SHOW_LABELS = true;
const DEFAULT_TEXT_OVERFLOW = "truncate";
const DEFAULT_TRUNCATION_MODE = "middle";
const DEFAULT_TOOLTIP_MAX_LINES = 1;
const DEFAULT_INTEGRATION_TEXT_OVERFLOW = "expand";
var TestIds = /* @__PURE__ */ ((TestIds2) => {
  TestIds2["Marker"] = "marker";
  TestIds2["Track"] = "track";
  TestIds2["MarkerGroup"] = "marker-group";
  TestIds2["AnnotationIndicator"] = "annotation-indicator";
  return TestIds2;
})(TestIds || {});
const MARKER_ICON_OFFSET = 2;
const MARKER_LABEL_PADDING = 6;
const FOREIGN_OBJECT_THRESHOLD = 8;
const MARKER_BORDER_RADIUS = 2;
const MARKER_CONTENT_TO_BORDER_PADDING = 4;
const MARGIN_TOP_STROKE = 1;
