// @dynatrace/strato-components/charts/gauge/components/GaugeLabel/GaugeLabel.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var GaugeLabel_exports = {};
__export(GaugeLabel_exports, {
  GaugeLabel: () => GaugeLabel
});
module.exports = __toCommonJS(GaugeLabel_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_GaugeLabel_css = require("./GaugeLabel.sty.js");
var import_GaugeResponsiveLabel = require("../GaugeResponsiveLabel/GaugeResponsiveLabel.js");
const GaugeLabel = ({
  value,
  autoHide,
  responsive,
  prefixIcon,
  ...textProps
}) => {
  const isElement = (0, import_react.isValidElement)(value);
  if (isElement && responsive) {
    return value;
  }
  if (responsive) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_GaugeResponsiveLabel.GaugeResponsiveLabel,
      {
        text: value?.toString() ?? "",
        id: textProps.id,
        prefixIcon
      }
    );
  }
  if (isElement) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "foreignObject",
      {
        x: textProps.x,
        y: textProps.y,
        fontSize: textProps.fontSize,
        className: import_GaugeLabel_css.elementLabelWrapper,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          "div",
          {
            className: (0, import_clsx.default)(
              (0, import_GaugeLabel_css.foreignObjectContainer)({
                dominantBaseline: textProps.dominantBaseline
              }),
              { [import_GaugeLabel_css.textContainer]: autoHide }
            ),
            children: value
          }
        )
      }
    );
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "text",
    {
      textAnchor: "middle",
      dominantBaseline: "hanging",
      ...textProps,
      className: (0, import_clsx.default)({ [import_GaugeLabel_css.textContainer]: autoHide }),
      children: value
    }
  );
};
