// @dynatrace/strato-components/charts/timeseries/messages/tooltips-text.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var tooltips_text_exports = {};
__export(tooltips_text_exports, {
  TIMESERIES_MESSAGES: () => TIMESERIES_MESSAGES
});
module.exports = __toCommonJS(tooltips_text_exports);
var import_react_intl = require("react-intl");
const TIMESERIES_MESSAGES = (0, import_react_intl.defineMessages)({
  zoomActionArrowUp: {
    defaultMessage: "Up",
    id: "+MYvp4qc5Ti3ddoO",
    description: 'Key shortcut in a tooltip shown when hovering the "Zoom in" action in the time series chart toolbar'
  },
  zoomActionArrowDown: {
    defaultMessage: "Down",
    id: "TKRQN9BIBeD8VT1I",
    description: 'Key shortcut in a tooltip shown when hovering the "Zoom out" action in the time series chart toolbar'
  },
  zoomModeExplore: {
    id: "TTCLEQtF7ylLxvEx",
    defaultMessage: "Explore mode",
    description: 'Text in a tooltip shown when hovering the "Explore" mode in the time series chart toolbar'
  },
  zoomModeZoom: {
    id: "KlA+bTF6CnwSnbCJ",
    defaultMessage: "Zoom mode",
    description: 'Text in a tooltip shown when hovering the "Zoom" mode in the timeseries chart toolbar'
  },
  zoomModeZoomX: {
    id: "VL8nEatHCgCkS4Yw",
    defaultMessage: "X-axis zoom",
    description: 'Text in a tooltip shown when hovering the "X-axis zoom" mode in the time series chart toolbar'
  },
  zoomModePan: {
    id: "ntRg2/q/0hsoJ9+r",
    defaultMessage: "Pan mode",
    description: 'Text in a tooltip shown when hovering the "Pan" mode in the time series chart toolbar'
  },
  zoomActionZoomIn: {
    id: "iN3ZVFNmBN0hzG0I",
    defaultMessage: "Zoom in",
    description: 'Text in a tooltip shown when hovering the "Zoom in" action in the time series chart toolbar'
  },
  zoomActionZoomOut: {
    id: "arDem4VUr6iNJ1A3",
    defaultMessage: "Zoom out",
    description: 'Text in a tooltip shown when hovering the "Zoom out" action in the time series chart toolbar'
  },
  zoomActionReset: {
    id: "2zjArS0xoR2Fy4/o",
    defaultMessage: "Reset",
    description: 'Text in a tooltip shown when hovering the "Reset" action in the time series chart toolbar'
  },
  downloadAsCsv: {
    id: "u5xmuzxNDovVbbQX",
    defaultMessage: "Download as CSV",
    description: 'Text in a tooltip shown when hovering the "Download data" button in the time series chart toolbar'
  },
  datapoints: {
    id: "1qd7igf1SoPYptNg",
    defaultMessage: "{count} data points",
    description: "Text in the tooltip header showing the number of selected data points"
  },
  noDatapoints: {
    id: "6GOnX5pixiNcVALE",
    defaultMessage: "No data points",
    description: "Text in the tooltip header showing that are no data points selected"
  },
  seriesDatapoints: {
    id: "nDzGVBRq+V8ulNuu",
    defaultMessage: "{count, plural, one {# data point} other {# data points}} across {seriesCount} series",
    description: "Text in the tooltip header showing the number of selected data points in multiple series"
  },
  zoomToSelection: {
    id: "af0kT/X4xYo3usY+",
    defaultMessage: "Zoom to selection",
    description: "Text displayed in the tooltip as the first action when zoom is enabled and a select interaction is performed on the Timeseries chart."
  }
});
