// @dynatrace/strato-components/content/microguide/contexts/MicroguideContext.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MicroguideContext_exports = {};
__export(MicroguideContext_exports, {
  MicroguideContext: () => MicroguideContext
});
module.exports = __toCommonJS(MicroguideContext_exports);
var import_react = require("react");
const DEFAULT_CONTEXT = {
  currentStep: 1,
  totalSteps: 1,
  isNext: true,
  onClose: () => {
  },
  setHasTitle: () => {
  },
  setHasContent: () => {
  },
  closeButtonRef: {
    current: null
  },
  closedOnKeydown: {
    current: false
  }
};
const MicroguideContext = (0, import_react.createContext)(DEFAULT_CONTEXT);
