// @dynatrace/strato-components/esm/charts/core/components/shared-chart-interactions/components/SynchronizedSelectionArea.js
import { jsx } from "react/jsx-runtime";
import { isUndefined } from "lodash-es";
import { useGridPlotSize } from "../../chart-grid/hooks/useGridPlotSize.js";
import { SelectionBackdrop } from "../../selection-area/SelectionBackdrop.js";
import { useSynchronizedSelection } from "../hooks/useSynchronizedSelection.js";
import { containerSVG } from "../SharedChartInteractions.sty.js";
const SynchronizedSelectionArea = ({
  scale
}) => {
  const synchronizedSelectionArea = useSynchronizedSelection();
  const { height } = useGridPlotSize();
  const { startX, endX, resetSelectionState } = synchronizedSelectionArea;
  if (!synchronizedSelectionArea || isUndefined(startX) || isUndefined(endX)) {
    return null;
  }
  const scaledStartX = scale(startX);
  const scaledDndX = scale(endX);
  const handleOnPointerDown = (event) => {
    event.preventDefault();
    resetSelectionState?.();
  };
  return /* @__PURE__ */ jsx("svg", { className: containerSVG, onPointerDown: handleOnPointerDown, children: /* @__PURE__ */ jsx(
    SelectionBackdrop,
    {
      focusable: false,
      interactable: false,
      startX: scaledStartX,
      endX: scaledDndX,
      height,
      isSynchronizedSelection: true
    }
  ) });
};
SynchronizedSelectionArea["displayName"] = "SynchronizedSelectionArea";
export {
  SynchronizedSelectionArea
};
