// @dynatrace/strato-components/esm/charts/core/components/chart-states/ChartStateHandler.js
import { jsx, jsxs } from "react/jsx-runtime";
import { clsx } from "clsx";
import { ErrorBoundary } from "react-error-boundary";
import { container } from "./ChartState.sty.js";
import { EmptyState } from "./EmptyState.js";
import { ErrorState } from "./ErrorState.js";
import { useBehavioralTrackingProps } from "../../../../core/hooks/useBehavioralTrackingProps.js";
import { LoaderOverlay } from "../loader-overlay/LoaderOverlay.js";
const getContent = (props) => {
  const { empty, height, emptyState, errorState, children, isLoading } = props;
  if (empty) {
    return isLoading ? /* @__PURE__ */ jsx(LoaderOverlay, {}) : /* @__PURE__ */ jsx(EmptyState, { height, children: emptyState });
  }
  return /* @__PURE__ */ jsxs(
    ErrorBoundary,
    {
      FallbackComponent: ErrorState({
        height,
        children: errorState
      }),
      children: [
        children,
        isLoading && /* @__PURE__ */ jsx(LoaderOverlay, {})
      ]
    }
  );
};
function ChartStateHandler(props) {
  const {
    height,
    empty,
    emptyState,
    errorState,
    isLoading,
    style,
    className,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    ...rest
  } = props;
  const content = getContent(props);
  const [behavioralTrackingProps, otherProps] = useBehavioralTrackingProps(rest);
  return /* @__PURE__ */ jsx(
    "div",
    {
      ...behavioralTrackingProps,
      className: clsx(container, className),
      style: {
        height,
        filter: isLoading ? "opacity(85%)" : "none",
        ...style
      },
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...otherProps,
      children: content
    }
  );
}
export {
  ChartStateHandler
};
