// @dynatrace/strato-components/esm/charts/core/components/chart-states/ErrorState.js
import { jsx, jsxs } from "react/jsx-runtime";
import { FormattedMessage } from "react-intl";
import { CriticalIcon } from "@dynatrace/strato-icons";
import { errorText, wrapper } from "./ChartState.sty.js";
import { Flex } from "../../../../layouts/flex/Flex.js";
import { Text } from "../../../../typography/text/Text.js";
const ErrorStateTextWrapper = ({ height, children }) => /* @__PURE__ */ jsx(
  Flex,
  {
    justifyContent: "center",
    alignItems: "center",
    height,
    width: "100%",
    children: /* @__PURE__ */ jsx(Text, { className: errorText, textStyle: "base", children })
  }
);
const ErrorState = ({ height, children }) => (fallbackProps) => {
  let errorNode;
  if (children && typeof children === "function") {
    errorNode = children(fallbackProps.error.message);
  } else {
    const errorMsg = fallbackProps.error.message ? `: ${fallbackProps.error.message}` : "";
    errorNode = children || /* @__PURE__ */ jsxs(Flex, { className: wrapper, children: [
      /* @__PURE__ */ jsx(CriticalIcon, {}),
      /* @__PURE__ */ jsx(
        FormattedMessage,
        {
          id: "9SDB/Tz9eGlVpihz",
          defaultMessage: "An error occurred",
          description: "Message to display when there is an error"
        }
      ),
      errorMsg
    ] });
  }
  return /* @__PURE__ */ jsx(ErrorStateTextWrapper, { height, children: errorNode });
};
ErrorState["displayName"] = "ErrorState";
export {
  ErrorState
};
