// @dynatrace/strato-components/charts/core/components/annotations/components/tooltip/AnnotationsDefaultTooltipContent.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var AnnotationsDefaultTooltipContent_exports = {};
__export(AnnotationsDefaultTooltipContent_exports, {
  AnnotationsDefaultTooltipContent: () => AnnotationsDefaultTooltipContent
});
module.exports = __toCommonJS(AnnotationsDefaultTooltipContent_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_units = require("@dynatrace-sdk/units");
var import_useAnnotationsTooltipFormatter = require("./hooks/useAnnotationsTooltipFormatter.js");
var import_get_symbol = require("./utils/get-symbol.js");
var import_Divider = require("../../../../../../layouts/divider/Divider.js");
var import_Footer = require("../../../overlay-tooltip/components/Footer.js");
var import_Header = require("../../../overlay-tooltip/components/header/Header.js");
var import_SeriesItem = require("../../../overlay-tooltip/components/series/SeriesItem.js");
var import_ExternalConfigContext = require("../../../overlay-tooltip/contexts/ExternalConfigContext.js");
var import_messages = require("../../../overlay-tooltip/messages.js");
var import_selectors = require("../../../overlay-tooltip/selectors.js");
var import_tooltip_css = require("../../../overlay-tooltip/tooltip.sty.js");
var import_useAnnotationsActions = require("../../hooks/useAnnotationsActions.js");
var import_serialize_marker_range = require("../../utils/serialize-marker-range.js");
const BASE_ACTIONS_CONFIG = /* @__PURE__ */ new Map([
  ["name", { field: "valueLabel" }]
]);
const buildExternalConfig = (truncationMode) => ({
  seriesHeader: null,
  displayHeaderDivider: false,
  symbolPosition: "left",
  truncationMode,
  defaultActionsConfig: BASE_ACTIONS_CONFIG,
  variant: "single"
});
const X_VALUE_LABEL = "X-value";
const AnnotationsDefaultTooltipContent = (props) => {
  const { data, tooltipType, truncationMode, maxLines } = props;
  const formatter = (0, import_useAnnotationsTooltipFormatter.useAnnotationsTooltipFormatter)();
  const annotationsActions = (0, import_useAnnotationsActions.useAnnotationsActions)();
  const config = buildExternalConfig(truncationMode);
  const intl = (0, import_react_intl.useIntl)();
  const unpinnedPlaceholderText = intl.formatMessage(
    import_messages.OVERLAY_TOOLTIP_MESSAGES.unpinnedPlaceholderMessage
  );
  const pinned = (0, import_selectors.useSelector)((state) => state.pinned);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_ExternalConfigContext.ExternalConfigContext.Provider, { value: config, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: import_tooltip_css.annotationsGroupsScrollContainer, children: data.map((annotationGroup, groupIndex) => {
      const [from, to] = (0, import_serialize_marker_range.deserializeMarkerRange)(annotationGroup.key);
      const isRange = to > from;
      let header1;
      let showRangeDuration = false;
      let duration;
      if (tooltipType === "numerical" && isRange) {
        header1 = [formatter(from), formatter(to)];
        duration = String(Math.abs(to - from));
      } else if (tooltipType === "numerical") {
        header1 = formatter(from);
      } else if (isRange) {
        header1 = [new Date(from), new Date(to)];
        showRangeDuration = true;
      } else {
        header1 = (0, import_units.formatDate)(new Date(from), {
          day: "2-digit",
          month: "short",
          year: "numeric",
          hour: "2-digit",
          minute: "2-digit"
        });
      }
      return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_react.Fragment, { children: [
        groupIndex > 0 && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Divider.Divider, { orientation: "horizontal", className: import_tooltip_css.divider }),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_Header.Header,
          {
            header1,
            label: X_VALUE_LABEL,
            showRangeDuration1: showRangeDuration,
            duration1: duration,
            maxLines
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: import_tooltip_css.seriesGridContainer, children: annotationGroup.markers.map((marker) => {
          const icon = (0, import_get_symbol.getSymbol)(marker);
          const {
            id,
            symbol,
            title: markerTitle,
            description: markerDescription,
            start,
            end,
            color,
            priority,
            hidden: markerHidden,
            ...extraFields
          } = marker;
          const payload = {
            ...extraFields,
            start,
            end,
            title: markerTitle,
            description: markerDescription,
            symbol
          };
          const actions = annotationsActions?.(payload);
          return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_SeriesItem.SeriesItem,
            {
              name: markerTitle ?? "",
              description: markerDescription,
              value: "",
              icon,
              multiLine: maxLines > 1,
              maxLines,
              actions: actions ? () => actions : void 0
            },
            id
          );
        }) })
      ] }, annotationGroup.key);
    }) }),
    !pinned && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Footer.Footer,
      {
        actions: [],
        unpinnedPlaceholderText
      }
    )
  ] });
};
AnnotationsDefaultTooltipContent.displayName = "AnnotationsDefaultTooltipContent";
