// @dynatrace/strato-components/charts/core/components/annotations/components/tooltip/AnnotationsOverlayTooltip.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var AnnotationsOverlayTooltip_exports = {};
__export(AnnotationsOverlayTooltip_exports, {
  AnnotationsOverlayTooltip: () => AnnotationsOverlayTooltip
});
module.exports = __toCommonJS(AnnotationsOverlayTooltip_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_AnnotationsCustomizedTooltip = require("./AnnotationsCustomizedTooltip.js");
var import_AnnotationsDefaultTooltipContent = require("./AnnotationsDefaultTooltipContent.js");
var import_useAnnotationsTooltipConfig = require("./hooks/useAnnotationsTooltipConfig.js");
var import_OverlayTooltip = require("../../../overlay-tooltip/OverlayTooltip.js");
var import_useTooltipReducer = require("../../../overlay-tooltip/useTooltipReducer.js");
var import_utils = require("../../../overlay-tooltip/utils.js");
var import_selectors = require("../../hooks/store/selectors.js");
var import_state_setters = require("../../hooks/store/state-setters.js");
var import_useMarkersInteractions = require("../../hooks/useMarkersInteractions.js");
var import_AnnotationsDataInitialState = require("../../providers/AnnotationsDataInitialState.js");
const buildSentinelData = () => ({
  geometry: "rect",
  series: [
    {
      name: "",
      value: "",
      color: ""
    }
  ]
});
const AnnotationsOverlayTooltip = (props) => {
  const { tooltipType, truncationMode = "end", maxLines = 1 } = props;
  const annotationsState = (0, import_selectors.useAnnotationsTooltipState)();
  const { pinned, data, visibility, position } = annotationsState;
  const dispatch = (0, import_useTooltipReducer.useTooltipReducer)();
  const updateAnnotationsTooltipState = (0, import_state_setters.useSetStateTooltip)();
  const updateAnnotationsSelectionState = (0, import_state_setters.useSetStateSelection)();
  const onMarkersInteraction = (0, import_useMarkersInteractions.useMarkersInteractions)();
  const config = (0, import_useAnnotationsTooltipConfig.useAnnotationsTooltipConfig)();
  const hidden = config?.hidden === true;
  const hasData = !!data && data.length > 0 && visibility;
  const prevPinnedRef = (0, import_react.useRef)(false);
  (0, import_react.useEffect)(() => {
    const wasPinned = prevPinnedRef.current;
    prevPinnedRef.current = pinned;
    if (wasPinned && !pinned) {
      dispatch({
        type: "SET_PINNED",
        payload: {
          pinned: false,
          data: (0, import_utils.emptyData)()
        }
      })();
    }
    if (hidden || !hasData || !position) {
      dispatch({ type: "SET_OUT_BOUNDS" })();
      return;
    }
    if (pinned) {
      dispatch({
        type: "SET_PINNED",
        payload: {
          pinned: true,
          absoluteX: position.x,
          absoluteY: position.y,
          data: buildSentinelData()
        }
      })();
    } else {
      dispatch({
        type: "SET_POSITION",
        payload: {
          absoluteX: position.x,
          absoluteY: position.y,
          data: buildSentinelData()
        }
      })();
    }
  }, [hidden, hasData, pinned, position?.x, position?.y]);
  (0, import_react.useEffect)(() => {
    return () => {
      dispatch({ type: "RESET_TOOLTIP" })();
      updateAnnotationsTooltipState(import_AnnotationsDataInitialState.initialState.tooltip);
      updateAnnotationsSelectionState(import_AnnotationsDataInitialState.initialState.selection);
      onMarkersInteraction?.(import_AnnotationsDataInitialState.initialState.selection);
    };
  }, []);
  (0, import_react.useEffect)(() => {
    const handleEscape = (event) => {
      if (event.key !== "Escape") {
        return;
      }
      updateAnnotationsTooltipState(import_AnnotationsDataInitialState.initialState.tooltip);
      updateAnnotationsSelectionState(import_AnnotationsDataInitialState.initialState.selection);
      onMarkersInteraction?.(import_AnnotationsDataInitialState.initialState.selection);
    };
    window.addEventListener("keyup", handleEscape);
    return () => {
      window.removeEventListener("keyup", handleEscape);
    };
  }, []);
  if (hidden || !hasData) {
    return null;
  }
  const legacyTemplate = config?.template ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_AnnotationsCustomizedTooltip.AnnotationsCustomizedTooltip,
    {
      tooltipTemplate: config.template,
      data
    }
  ) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_AnnotationsDefaultTooltipContent.AnnotationsDefaultTooltipContent,
    {
      data,
      tooltipType,
      truncationMode,
      maxLines
    }
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayTooltip.OverlayTooltip, { legacyTemplate });
};
AnnotationsOverlayTooltip.displayName = "AnnotationsOverlayTooltip";
