// @dynatrace/strato-components/charts/histogram/components/annotations-histogram/HistogramAnnotationsArea.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var HistogramAnnotationsArea_exports = {};
__export(HistogramAnnotationsArea_exports, {
  HistogramAnnotationsArea: () => HistogramAnnotationsArea
});
module.exports = __toCommonJS(HistogramAnnotationsArea_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_constants = require("./constants.js");
var import_useBehavioralTrackingProps = require("../../../../core/hooks/useBehavioralTrackingProps.js");
var import_AnnotationsNumericalDisplay = require("../../../core/components/annotations/components/displays/AnnotationsNumericalDisplay.js");
var import_AnnotationsScrollWrapper = require("../../../core/components/annotations/components/overlays/AnnotationsScrollWrapper.js");
var import_constants2 = require("../../../core/components/annotations/constants.js");
var import_ChartGridRoot = require("../../../core/components/chart-grid/ChartGridRoot.js");
var import_useGridPlotSize = require("../../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_histogram_grid_area_rules = require("../../constants/histogram-grid-area-rules.js");
var import_state_setters = require("../../hooks/store/state-setters.js");
var import_useHistogramAnnotationsConfig = require("../../hooks/useHistogramAnnotationsConfig.js");
var import_useHistogramXScale = require("../../hooks/useHistogramXScale.js");
var import_useXAxisConfig = require("../../hooks/useXAxisConfig.js");
const COMPONENT_NAME = "AnnotationsChart";
const HistogramAnnotationsArea = () => {
  const { xScale } = (0, import_useHistogramXScale.useHistogramXScale)();
  const config = (0, import_useHistogramAnnotationsConfig.useHistogramAnnotationsConfig)();
  const {
    tracks,
    showLabels = import_constants2.DEFAULT_SHOW_LABELS,
    annotationsActions,
    tooltip
  } = config;
  const [behaviorTrackingProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(config);
  const setState = (0, import_state_setters.useSetState)();
  const { width: trackWidth } = (0, import_useGridPlotSize.useGridPlotSize)();
  const handleUpdateIndicators = (event) => {
    setState((prev) => ({
      ...prev,
      annotations: {
        ...prev.annotations,
        ...event
      }
    }));
  };
  const { formatter = (value) => `${(0, import_lodash_es.toInteger)(value)}` } = (0, import_useXAxisConfig.useXAxisConfig)();
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ChartGridRoot.ChartGrid.GridArea,
    {
      paddingBottom: import_constants.ANNOTATIONS_PADDING_BOTTOM,
      gridArea: import_histogram_grid_area_rules.AREAS.annotations,
      children: ({ availableWidth, availableHeight }) => {
        const labelOffset = availableWidth - trackWidth;
        return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_AnnotationsScrollWrapper.AnnotationsScrollWrapper,
          {
            height: availableHeight,
            width: availableWidth + import_constants2.SCROLLBAR_THRESHOLD,
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_AnnotationsNumericalDisplay.AnnotationsNumericalDisplay,
              {
                tracks,
                width: availableWidth,
                scale: xScale,
                showLabels,
                labelsOffset: labelOffset,
                annotationsActions: (annotation) => annotation && annotationsActions?.(annotation),
                markerGroupHover: handleUpdateIndicators,
                formatter,
                textOverflow: import_constants2.DEFAULT_INTEGRATION_TEXT_OVERFLOW,
                tooltipConfig: tooltip,
                ...behaviorTrackingProps,
                "data-dt-component": COMPONENT_NAME
              }
            )
          }
        );
      }
    }
  );
};
HistogramAnnotationsArea["displayName"] = "HistogramAnnotationsArea";
