// @dynatrace/strato-components/charts/timeseries/utils/convert-to-timeseries/convert-to-timeseries-band.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var convert_to_timeseries_band_exports = {};
__export(convert_to_timeseries_band_exports, {
  convertToTimeseriesBand: () => convertToTimeseriesBand
});
module.exports = __toCommonJS(convert_to_timeseries_band_exports);
var import_create_timeseries_from_record = require("./utils/create-timeseries-from-record.js");
const upperSuffix = ["max", "upper"];
const lowerSuffix = ["min", "lower"];
function valueDimensionEndsWith(timeseries, suffixList) {
  return suffixList.some(
    (suffix) => timeseries.userData.valueDimension.endsWith(suffix)
  );
}
function getDimensionsWithoutLast(name) {
  if (typeof name === "string") {
    return "band";
  }
  const timeseriesName = name.slice(0, -1);
  return timeseriesName.length > 0 ? timeseriesName : "band";
}
function convertToTimeseriesBand(record, types) {
  const timeseries = (0, import_create_timeseries_from_record.createTimeseriesFromRecord)(record, types);
  const upper = timeseries.find(
    (timeseries2) => valueDimensionEndsWith(timeseries2, upperSuffix)
  );
  const lower = timeseries.find(
    (timeseries2) => valueDimensionEndsWith(timeseries2, lowerSuffix)
  );
  const rest = timeseries.filter(
    (timeseries2) => timeseries2 !== lower && timeseries2 !== upper
  );
  if (upper && lower && upper.datapoints.length === lower.datapoints.length) {
    const timeseriesBand = {
      name: getDimensionsWithoutLast(upper.name),
      datapoints: upper.datapoints.map(
        ({ value, ...datapoint }, index) => ({
          ...datapoint,
          y0: lower?.datapoints[index].value,
          y1: value
        })
      ),
      userData: {
        ...upper.userData,
        lowerValueDimension: lower.userData.valueDimension,
        upperValueDimension: upper.userData.valueDimension
      }
    };
    return {
      timeseriesBand,
      timeseries: rest,
      timeseriesOfBand: [upper, lower]
    };
  }
}
