// @dynatrace/strato-components/charts/timeseries/utils/convert-to-timeseries/convert-to-timeseries.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var convert_to_timeseries_exports = {};
__export(convert_to_timeseries_exports, {
  convertQueryResultToTimeseries: () => convertQueryResultToTimeseries,
  convertToTimeseries: () => convertToTimeseries
});
module.exports = __toCommonJS(convert_to_timeseries_exports);
var import_units = require("@dynatrace-sdk/units");
var import_create_name = require("./create-name.js");
var import_create_timeseries_from_record = require("./utils/create-timeseries-from-record.js");
var import_extract_start_end_dates = require("./utils/extract-start-end-dates.js");
var import_extractValue = require("./utils/extractValue.js");
var import_get_dimensions = require("./utils/get-dimensions.js");
var import_get_metadata_metrics_unit_for_field_name = require("./utils/get-metadata-metrics-unit-for-field-name.js");
var import_getFieldsArrayForRecordIndex = require("./utils/getFieldsArrayForRecordIndex.js");
var import_unit_mapping = require("./utils/unit-mapping.js");
function createKeyFromTimeseries(name) {
  if (typeof name === "string") {
    return name;
  }
  const dimensionKeys = [...name];
  const last = dimensionKeys.pop();
  const baseKey = dimensionKeys.join("-");
  return baseKey ? `${baseKey}_${last}` : last ?? "";
}
function convertQueryResultToTimeseries(queryResult) {
  if (queryResult.records === null) {
    return [];
  }
  return convertToTimeseries(
    queryResult.records,
    queryResult.types,
    void 0,
    queryResult.metadata
  );
}
function convertToTimeseries(records, fieldTypes, hiddenFieldIds, metadata) {
  const timeseriesMap = /* @__PURE__ */ new Map();
  const valueDimensionSet = /* @__PURE__ */ new Set();
  for (const [recordIndex, record] of records.entries()) {
    if (record === null) {
      continue;
    }
    const types = (0, import_getFieldsArrayForRecordIndex.getTypesForRecordIndex)(fieldTypes, recordIndex);
    if (!types || !record) {
      break;
    }
    const timeseriesArray = (0, import_create_timeseries_from_record.createTimeseriesFromRecord)(record, types);
    if (timeseriesArray.length > 0) {
      timeseriesArray.forEach((timeseries2) => {
        const valueDimension = timeseries2.userData.valueDimension;
        valueDimensionSet.add(valueDimension);
        if (!hiddenFieldIds?.includes(valueDimension)) {
          const unit = (0, import_get_metadata_metrics_unit_for_field_name._getMetadataMetricsUnitForFieldName)(
            metadata,
            valueDimension
          );
          const newTimeseries = {
            ...timeseries2,
            unit: unit !== void 0 ? (0, import_unit_mapping.mapMetricUnitToSdk)(unit) : void 0
          };
          const mapKey = createKeyFromTimeseries(timeseries2.name);
          const uniqueMapKeyAcrossRecords = JSON.stringify([
            mapKey,
            recordIndex
          ]);
          timeseriesMap.set(uniqueMapKeyAcrossRecords, newTimeseries);
        }
      });
      continue;
    }
    const fieldsArray = (0, import_getFieldsArrayForRecordIndex.getFieldsArrayForRecordIndex)(fieldTypes, recordIndex);
    let dimensions = (0, import_get_dimensions.getDimensions)(record, fieldsArray, false);
    if (hiddenFieldIds && hiddenFieldIds.length) {
      dimensions = dimensions.filter((dimension) => {
        return !hiddenFieldIds.includes(dimension.field.name);
      });
    }
    const [start, end] = (0, import_extract_start_end_dates.extractStartEndDates)(record, fieldsArray);
    const valueDimensions = (0, import_get_dimensions.getValueDimensions)(fieldsArray, false);
    for (const valueDimension of valueDimensions ?? []) {
      const name = (0, import_create_name.createName)(dimensions, valueDimension.name);
      valueDimensionSet.add(valueDimension.name);
      if (name.length === 0) {
        throw new Error(
          `Could not calculate key for record ${JSON.stringify(record)}`
        );
      }
      if (hiddenFieldIds?.includes(valueDimension.name)) {
        continue;
      }
      const timeseries2 = ensureMapEntry(
        timeseriesMap,
        name,
        dimensions,
        valueDimension.name
      );
      const rawValue = (0, import_extractValue.extractValue)(
        record,
        valueDimension.name
      );
      const value = typeof rawValue === "string" ? parseFloat(rawValue) : rawValue;
      if (valueDimension.type === "duration") {
        timeseries2.unit = import_units.units.time.nanosecond;
      }
      if (value !== void 0) {
        timeseries2.datapoints.push({ start, end, value });
      }
    }
  }
  const timeseries = Array.from(timeseriesMap.values()).filter(
    (timeseries2) => timeseries2.datapoints.length > 0
  );
  timeseries.forEach((t) => {
    t.datapoints.sort((a, b) => a.start.getTime() - b.start.getTime());
    if (valueDimensionSet.size === 1 && Array.isArray(t.name) && t.name.length > 1) {
      t.name.pop();
    }
  });
  return timeseries;
}
function ensureMapEntry(timeseriesMap, name, dimensions, valueDimension) {
  const key = createKeyFromTimeseries(name);
  if (key !== void 0 && key !== null && !timeseriesMap.has(key)) {
    const timeseries = {
      datapoints: [],
      name,
      userData: {
        dimensions,
        valueDimension
      }
    };
    timeseriesMap.set(key, timeseries);
    return timeseries;
  }
  return timeseriesMap.get(key);
}
