// @dynatrace/strato-components/charts/timeseries/components/toolbar/ExploreButton.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ExploreButton_exports = {};
__export(ExploreButton_exports, {
  ExploreButton: () => ExploreButton
});
module.exports = __toCommonJS(ExploreButton_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_useGridPlotSize = require("../../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_types = require("../../../core/components/interactions-overlay/types.js");
var import_calculate_initial_selection_area_positions = require("../../../core/components/interactions-overlay/utils/calculate-initial-selection-area-positions.js");
var import_explore_reducers = require("../../../core/components/interactions-state-helpers/reducers/explore-reducers.js");
var import_mode_reducers = require("../../../core/components/interactions-state-helpers/reducers/mode-reducers.js");
var import_useSynchronizedSelection = require("../../../core/components/shared-chart-interactions/hooks/useSynchronizedSelection.js");
var import_ChartToolbarRenderer = require("../../../core/components/toolbar/ChartToolbarRenderer.js");
var import_useToolbarContext = require("../../../core/components/toolbar/hooks/useToolbarContext.js");
var import_useA11yToolbarTabIndex = require("../../../core/hooks/useA11yToolbarTabIndex.js");
var import_get_percentage_from_pixel = require("../../../core/utils/get-percentage-from-pixel.js");
var import_selectors = require("../../hooks/store/selectors.js");
var import_useChartInteractions = require("../../hooks/useChartInteractions.js");
var import_useTimeseriesChartId = require("../../hooks/useTimeseriesChartId.js");
var import_useTimeseriesOverlayRef = require("../../hooks/useTimeseriesOverlayRef.js");
var import_useXScale = require("../../hooks/useXScale.js");
var import_tooltips_text = require("../../messages/tooltips-text.js");
const ExploreButton = () => {
  const intl = (0, import_react_intl.useIntl)();
  const setState = (0, import_selectors.useSetState)();
  const tabIndex = (0, import_useA11yToolbarTabIndex.useA11yToolbarTabIndex)(import_types.INTERACTION_MODE.Explore);
  const { pan, zoom } = (0, import_useChartInteractions.useChartInteractions)();
  const overlayRef = (0, import_useTimeseriesOverlayRef.useTimeseriesOverlayRef)();
  const { updateLastItemSelected } = (0, import_useToolbarContext.useToolbarContext)();
  const { currentScale: xScale } = (0, import_useXScale.useXScales)();
  const chartId = (0, import_useTimeseriesChartId.useTimeseriesChartId)();
  const selectState = (0, import_selectors.useSelectState)();
  const { width } = (0, import_useGridPlotSize.useGridPlotSize)();
  const { updateSelectionState, resetSelectionState } = (0, import_useSynchronizedSelection.useSynchronizedSelection)();
  const interactionMode = (0, import_selectors.useInteractionModeState)();
  const isSelected = interactionMode === import_types.INTERACTION_MODE.Explore;
  const modePlacement = pan?.enabled || zoom?.enabled ? "menu" : "auto";
  const createInitialSelectionArea = () => {
    if (!(0, import_lodash_es.isNil)(overlayRef.current)) {
      const bounds = overlayRef.current.getBoundingClientRect();
      const [start, end] = (0, import_calculate_initial_selection_area_positions.calculateInitialSelectionAreaPositions)(bounds);
      setState(
        (0, import_explore_reducers.reduceExploreSelect)({
          start,
          end,
          finished: true,
          accessible: true,
          variant: "select",
          isKeyboardCreated: true,
          selectionPercentage: (0, import_get_percentage_from_pixel.getPercentageFromPixel)([start, end], width)
        })
      );
      updateSelectionState?.({
        selection: [start, end],
        chartId,
        scale: xScale
      });
    }
  };
  const handleKeyDown = (event) => {
    if (event.key === "Enter") {
      createInitialSelectionArea();
    }
  };
  const handleClick = () => {
    if (interactionMode !== "explore") {
      setState((0, import_mode_reducers.reduceInteractionMode)(import_types.INTERACTION_MODE.Explore));
    }
    if (selectState?.start && selectState?.end) {
      if (selectState?.isKeyboardCreated) {
        (0, import_mode_reducers.reduceSelectionAreaVariant)("select");
      } else {
        resetSelectionState?.();
        setState(import_explore_reducers.reduceExploreCancelAll);
      }
    }
    updateLastItemSelected(import_types.INTERACTION_MODE.Explore);
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ChartToolbarRenderer.ChartToolbarRenderer.Control,
    {
      prefixIcon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.CrosshairIcon, {}),
      tooltipText: intl.formatMessage(import_tooltips_text.TIMESERIES_MESSAGES.zoomModeExplore),
      tooltipKeyShortcut: "E",
      selected: isSelected,
      onKeyDown: handleKeyDown,
      onClick: handleClick,
      tabIndex,
      placement: modePlacement
    }
  );
};
ExploreButton["displayName"] = "ExploreButton";
