// @dynatrace/strato-components/charts/timeseries/components/toolbar/PanButton.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var PanButton_exports = {};
__export(PanButton_exports, {
  PanButton: () => PanButton
});
module.exports = __toCommonJS(PanButton_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_constants = require("../../../core/components/interactions-overlay/constants.js");
var import_types = require("../../../core/components/interactions-overlay/types.js");
var import_explore_reducers = require("../../../core/components/interactions-state-helpers/reducers/explore-reducers.js");
var import_mode_reducers = require("../../../core/components/interactions-state-helpers/reducers/mode-reducers.js");
var import_useSynchronizedSelection = require("../../../core/components/shared-chart-interactions/hooks/useSynchronizedSelection.js");
var import_ChartToolbarRenderer = require("../../../core/components/toolbar/ChartToolbarRenderer.js");
var import_useA11yToolbarTabIndex = require("../../../core/hooks/useA11yToolbarTabIndex.js");
var import_selectors = require("../../hooks/store/selectors.js");
var import_useChartInteractions = require("../../hooks/useChartInteractions.js");
var import_tooltips_text = require("../../messages/tooltips-text.js");
const PanButton = () => {
  const { zoomOutEnabled } = (0, import_selectors.useZoomState)();
  const panButtonRef = (0, import_react.useRef)(null);
  const tabIndex = (0, import_useA11yToolbarTabIndex.useA11yToolbarTabIndex)(import_types.INTERACTION_MODE.Pan);
  const { resetSelectionState } = (0, import_useSynchronizedSelection.useSynchronizedSelection)();
  const { pan } = (0, import_useChartInteractions.useChartInteractions)();
  const setState = (0, import_selectors.useSetState)();
  const intl = (0, import_react_intl.useIntl)();
  const isPanEnabled = pan?.enabled;
  const interactionMode = (0, import_selectors.useInteractionModeState)();
  const isSelected = interactionMode === import_types.INTERACTION_MODE.Pan;
  const handleClick = () => {
    setState(import_explore_reducers.reduceExploreCancelAll);
    resetSelectionState?.();
    setState((0, import_mode_reducers.reduceInteractionMode)(import_types.INTERACTION_MODE.Pan));
  };
  const handleKeyDown = (event) => {
    if (event.key === "Enter") {
      handleClick();
    }
  };
  return isPanEnabled ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ChartToolbarRenderer.ChartToolbarRenderer.Control,
    {
      ref: panButtonRef,
      prefixIcon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.PanHorizontalIcon, {}),
      tooltipText: intl.formatMessage(import_tooltips_text.TIMESERIES_MESSAGES.zoomModePan),
      tooltipKeyShortcut: import_constants.KeyboardControls.KEYBOARD_PAN_MODE_SHORTCUT_ALTERNATIVE,
      selected: isSelected,
      onKeyDown: handleKeyDown,
      onClick: handleClick,
      tabIndex,
      disabled: !zoomOutEnabled,
      placement: "menu"
    }
  ) : null;
};
PanButton["displayName"] = "PanButton";
