// @dynatrace/strato-components/charts/timeseries/components/toolbar/ResetButton.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ResetButton_exports = {};
__export(ResetButton_exports, {
  ResetButton: () => ResetButton
});
module.exports = __toCommonJS(ResetButton_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_useSynchronizedSelection = require("../../../core/components/shared-chart-interactions/hooks/useSynchronizedSelection.js");
var import_ChartToolbarRenderer = require("../../../core/components/toolbar/ChartToolbarRenderer.js");
var import_useA11yToolbarTabIndex = require("../../../core/hooks/useA11yToolbarTabIndex.js");
var import_zoom = require("../../../core/utils/zoom.js");
var import_selectors = require("../../hooks/store/selectors.js");
var import_useChartInteractions = require("../../hooks/useChartInteractions.js");
var import_useXScale = require("../../hooks/useXScale.js");
var import_tooltips_text = require("../../messages/tooltips-text.js");
var import_useResetInteractions = require("../overlays/hooks/useResetInteractions.js");
const ResetButton = () => {
  const { initialScale } = (0, import_useXScale.useXScales)();
  const intl = (0, import_react_intl.useIntl)();
  const resetButtonRef = (0, import_react.useRef)(null);
  const tabIndex = (0, import_useA11yToolbarTabIndex.useA11yToolbarTabIndex)("reset");
  const currentDomain = (0, import_selectors.useCurrentDomainState)();
  const { zoom, zoomX } = (0, import_useChartInteractions.useChartInteractions)();
  const resetInteractions = (0, import_useResetInteractions.useResetInteractions)();
  const { resetSelectionState } = (0, import_useSynchronizedSelection.useSynchronizedSelection)();
  const hasZoomInteractions = zoom?.enabled || zoomX?.enabled;
  const handleClick = () => {
    resetInteractions();
    resetSelectionState?.();
  };
  function handleZoomResetKeyboardMode(event) {
    if (event.code === "Enter") {
      resetInteractions();
    }
  }
  const isResettable = !(0, import_zoom.isSameDomain)(currentDomain, initialScale.domain());
  const onlyXZoomIsEnabled = zoomX?.enabled && !zoom?.enabled;
  return hasZoomInteractions ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ChartToolbarRenderer.ChartToolbarRenderer.Control,
    {
      ref: resetButtonRef,
      prefixIcon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ResetIcon, {}),
      tooltipText: intl.formatMessage(import_tooltips_text.TIMESERIES_MESSAGES.zoomActionReset),
      tooltipKeyShortcut: "R",
      testId: "chart-toolbar-zoom-reset",
      onKeyDown: handleZoomResetKeyboardMode,
      onClick: handleClick,
      tabIndex,
      disabled: !isResettable,
      placement: "toolbar",
      divider: onlyXZoomIsEnabled ? "left" : void 0,
      showWhenCollapsed: isResettable
    }
  ) : null;
};
ResetButton["displayName"] = "ResetButton";
