// @dynatrace/strato-components/charts/timeseries/components/toolbar/TimeseriesChartToolbar.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TimeseriesChartToolbar_exports = {};
__export(TimeseriesChartToolbar_exports, {
  TimeseriesChartToolbar: () => TimeseriesChartToolbar
});
module.exports = __toCommonJS(TimeseriesChartToolbar_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_react_intl = require("react-intl");
var import_DownloadDataButton = require("./DownloadDataButton.js");
var import_ExploreButton = require("./ExploreButton.js");
var import_PanButton = require("./PanButton.js");
var import_ResetButton = require("./ResetButton.js");
var import_ZoomInOutButtons = require("./ZoomInOutButtons.js");
var import_ZoomModeButton = require("./ZoomModeButton.js");
var import_useBehavioralTrackingProps = require("../../../../core/hooks/useBehavioralTrackingProps.js");
var import_ChartToolbarRenderer = require("../../../core/components/toolbar/ChartToolbarRenderer.js");
var import_get_menu_trigger_display_mode = require("../../../core/components/toolbar/utils/get-menu-trigger-display-mode.js");
var import_get_menu_trigger_icon = require("../../../core/components/toolbar/utils/get-menu-trigger-icon.js");
var import_constants = require("../../../core/constants.js");
var import_zoom = require("../../../core/utils/zoom.js");
var import_selectors = require("../../hooks/store/selectors.js");
var import_useChartInteractions = require("../../hooks/useChartInteractions.js");
var import_useChartToolbar = require("../../hooks/useChartToolbar.js");
var import_useTimeseriesRawData = require("../../hooks/useTimeseriesRawData.js");
var import_useTimeseriesSelectionAreaRef = require("../../hooks/useTimeseriesSelectionAreaRef.js");
var import_tooltips_text = require("../../messages/tooltips-text.js");
var import_build_and_download_csv = require("../../utils/download/build-and-download-csv.js");
const defaultToolbarPlacement = "top-right";
const TimeseriesChartToolbar = ({
  onMouseEnter,
  onMouseLeave,
  handlerRef
}) => {
  const selectionAreaRef = (0, import_useTimeseriesSelectionAreaRef.useTimeseriesSelectionAreaRef)();
  const config = (0, import_useChartToolbar.useChartToolbar)();
  const { zoom, zoomX, pan } = (0, import_useChartInteractions.useChartInteractions)();
  const intl = (0, import_react_intl.useIntl)();
  const isDownloadEnabled = config?.downloadData?.enabled ?? false;
  const isPanEnabled = pan?.enabled;
  const isZoomEnabled = zoom?.enabled;
  const isZoomXEnabled = zoomX?.enabled;
  const anyInteractionEnabled = isZoomEnabled || isZoomXEnabled || isPanEnabled || isDownloadEnabled || config?.intents;
  const mode = (0, import_selectors.useInteractionModeState)();
  const menuIcon = (0, import_get_menu_trigger_icon.getMenuTriggerIcon)(mode, isZoomEnabled);
  const menuLabel = !isPanEnabled && !isZoomEnabled ? intl.formatMessage(import_tooltips_text.TIMESERIES_MESSAGES.zoomModeExplore) : void 0;
  const rawData = (0, import_useTimeseriesRawData.useTimeseriesRawData)();
  const handleDownloadDataToolbar = () => (0, import_build_and_download_csv.buildAndDownloadCsv)(rawData);
  const menuDisplayMode = (0, import_get_menu_trigger_display_mode.getMenuTriggerDisplayMode)({
    isPanEnabled,
    isZoomEnabled,
    isDownloadEnabled
  });
  const onClose = () => selectionAreaRef ? selectionAreaRef.current?.focus() : void 0;
  const showDefaultToolbarDownloadButton = isDownloadEnabled && !config?.intents;
  const showMenuToolbarButton = isDownloadEnabled && !!config?.intents?.length;
  const behaviorTrackingProps = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(config ?? {})[0];
  return !(0, import_lodash_es.isUndefined)(config) && anyInteractionEnabled ? /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_ChartToolbarRenderer.ChartToolbarRenderer,
    {
      visible: false,
      handlerRef,
      visibleWhenMouseInBounds: true,
      collapsed: config.collapsed,
      draggable: true,
      placement: config.placement || defaultToolbarPlacement,
      initialA11yState: (0, import_zoom.getA11yToolbarConfig)({
        zoomX,
        pan,
        zoom,
        download: config.downloadData,
        intent: !!config.intents
      }),
      zIndex: import_constants.ChartOverlaysElevation.toolbar,
      onMouseEnter,
      onMouseLeave,
      menuIcon,
      menuTriggerLabel: menuLabel,
      menuDisplayMode,
      onClose,
      isDownloadEnabled: showMenuToolbarButton,
      intents: config?.intents,
      handleDownloadDataToolbar,
      ...behaviorTrackingProps,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ExploreButton.ExploreButton, {}),
          isZoomEnabled && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ZoomModeButton.ZoomModeButton, {}),
          isPanEnabled && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_PanButton.PanButton, {})
        ] }),
        isZoomEnabled && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ZoomInOutButtons.ZoomInOutButtons, {}),
        isZoomEnabled && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ResetButton.ResetButton, {}),
        showDefaultToolbarDownloadButton && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_DownloadDataButton.DownloadDataButton, {})
      ]
    }
  ) : null;
};
TimeseriesChartToolbar["displayName"] = "ToolbarChartToolbar";
