// @dynatrace/strato-components/esm/buttons/notify-button/NotifyButton.js
import { jsx, jsxs } from "react/jsx-runtime";
import { Children, forwardRef, useMemo } from "react";
import { isElement } from "react-is";
import { Content } from "./components/Content.js";
import { NotifyButtonTrigger } from "./components/Trigger.js";
import { NotifyButtonContext } from "./contexts/NotifyButtonContext.js";
import * as styles from "./NotifyButton.sty.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useId } from "../../core/hooks/useId.js";
import { useMergeRefs } from "../../core/hooks/useMergeRefs.js";
import { isDev } from "../../core/utils/is-node-environment.js";
import { mergeProps } from "../../core/utils/merge-props.js";
import { Flex } from "../../layouts/flex/Flex.js";
import { Content as NotificationSettingsContent } from "../../notifications/notification-settings/components/Content.js";
import { NotificationPermissionInfo } from "../../notifications/notification-settings/components/PermissionInfo.js";
import { NotificationToggleButton } from "../../notifications/notification-settings/components/ToggleButton.js";
import { NotificationProvider } from "../../notifications/notification-settings/NotificationSettings.js";
import { useOverlayWithTrigger } from "../../overlays/hooks/useOverlayWithTrigger.js";
import { Overlay } from "../../overlays/overlay/Overlay.js";
const COMPONENT_NAME = "NotifyButton";
function getContentSlot(children) {
  let content;
  Children.forEach(children, (child) => {
    if (isElement(child)) {
      switch (child.type) {
        case Content:
          content = child;
          break;
        default: {
          if (isDev()) {
            console.warn("Only Content is a valid child of the NotifyButton.");
          }
          break;
        }
      }
    } else if (child) {
      if (isDev()) {
        console.warn("The used child is not supported by the NotifyButton.");
      }
    }
  });
  return content;
}
const RootNotifyButton = forwardRef(
  (props, ref) => {
    const {
      children,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      className: consumerClassName,
      style: consumerStyle,
      disabled,
      size = "default",
      variant = "default",
      type,
      config,
      readOnly,
      showLabel = false,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
    const overlayAriaLabelledBy = useId();
    const {
      targetRef,
      triggerRef,
      overlayRef,
      overlayProps,
      overlayState,
      overlayTriggerProps
    } = useOverlayWithTrigger({
      placement: "bottom-start",
      placementFallbacks: [
        "bottom-start",
        "bottom-end",
        "top-start",
        "top-end"
      ],
      offset: 4,
      disabled: disabled || readOnly,
      "aria-labelledby": overlayAriaLabelledBy
    });
    const rootRef = useMergeRefs([ref, triggerRef, targetRef]);
    const contentSlot = useMemo(() => getContentSlot(children), [children]);
    const notifyButtonContextValue = useMemo(
      () => ({
        closeOverlay: overlayState.close,
        overlayAriaLabelledBy
      }),
      [overlayAriaLabelledBy, overlayState.close]
    );
    return /* @__PURE__ */ jsx(NotifyButtonContext.Provider, { value: notifyButtonContextValue, children: /* @__PURE__ */ jsxs(NotificationProvider, { type, config, children: [
      /* @__PURE__ */ jsx(
        NotifyButtonTrigger,
        {
          ref: rootRef,
          "data-testid": dataTestId,
          "data-dtrum-mask": dataDtrumMask,
          "data-dtrum-allow": dataDtrumAllow,
          className: consumerClassName,
          style: consumerStyle,
          disabled,
          size,
          variant,
          showLabel,
          triggerProps: overlayTriggerProps,
          "aria-disabled": readOnly,
          ...mergeProps(restProps, behavioralTrackingProps),
          "data-dt-component": COMPONENT_NAME
        }
      ),
      /* @__PURE__ */ jsx(
        Overlay,
        {
          ref: overlayRef,
          isOpen: overlayState.isOpen,
          className: styles.overlay,
          overlayProps,
          children: /* @__PURE__ */ jsxs(Flex, { flexDirection: "column", gap: 12, children: [
            contentSlot ?? /* @__PURE__ */ jsx(NotificationSettingsContent, {}),
            /* @__PURE__ */ jsxs(Flex, { alignItems: "center", children: [
              /* @__PURE__ */ jsx(NotificationToggleButton, {}),
              /* @__PURE__ */ jsx(NotificationPermissionInfo, {})
            ] })
          ] })
        }
      )
    ] }) });
  }
);
RootNotifyButton.displayName = COMPONENT_NAME;
const NotifyButton = Object.assign(RootNotifyButton, {
  Content
});
export {
  NotifyButton
};
