// @dynatrace/strato-components/charts/core/components/annotations/providers/AnnotationsDataInitialState.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var AnnotationsDataInitialState_exports = {};
__export(AnnotationsDataInitialState_exports, {
  initialSelectionState: () => initialSelectionState,
  initialState: () => initialState
});
module.exports = __toCommonJS(AnnotationsDataInitialState_exports);
const initialSelectionState = {
  hoveredGroup: void 0,
  pinnedGroup: void 0,
  hoveredItem: void 0,
  hoveredListItem: void 0,
  data: void 0,
  hoverGroupListKey: void 0
};
const initialTooltipState = {
  visibility: false,
  pinned: false,
  data: void 0,
  position: void 0,
  relativePosition: { x: 0, y: 0 }
};
const initialState = {
  tooltip: initialTooltipState,
  selection: initialSelectionState
};
