// @dynatrace/strato-components/charts/core/components/data-point/constants.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var constants_exports = {};
__export(constants_exports, {
  BOTTOM_PADDING_ADJUSTMENT: () => BOTTOM_PADDING_ADJUSTMENT,
  DATA_POINT_OUTER_RADIUS: () => DATA_POINT_OUTER_RADIUS,
  GROUP_DISTANCE_THRESHOLD: () => GROUP_DISTANCE_THRESHOLD,
  GROUP_PADDING: () => GROUP_PADDING,
  HALF_PIXEL_CORRECTION: () => HALF_PIXEL_CORRECTION,
  OVERLAY_PADDING: () => OVERLAY_PADDING,
  POINT_OVERLAY_PADDING: () => POINT_OVERLAY_PADDING
});
module.exports = __toCommonJS(constants_exports);
var import_constants = require("../../constants.js");
const DATA_POINT_OUTER_RADIUS = 6;
const GROUP_PADDING = DATA_POINT_OUTER_RADIUS - import_constants.DEFAULT_DATA_POINT_RADIUS;
const GROUP_DISTANCE_THRESHOLD = DATA_POINT_OUTER_RADIUS * 2;
const OVERLAY_PADDING = DATA_POINT_OUTER_RADIUS + 1;
const OUTER_LINE_SHAPE = 5;
const POINT_OVERLAY_PADDING = DATA_POINT_OUTER_RADIUS + OUTER_LINE_SHAPE;
const BOTTOM_PADDING_ADJUSTMENT = 1;
const HALF_PIXEL_CORRECTION = 1;
