// @dynatrace/strato-components/charts/core/components/data-point/DataPointCanvas.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DataPointCanvas_exports = {};
__export(DataPointCanvas_exports, {
  DataPointCanvas: () => DataPointCanvas
});
module.exports = __toCommonJS(DataPointCanvas_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
var import_constants = require("./constants.js");
var import_data_point = require("./utils/data-point.js");
var import_constants2 = require("../../constants.js");
var import_RoundRectCanvas = require("../canvas/components/RoundRectCanvas.js");
var import_SymbolCanvas = require("../canvas/components/SymbolCanvas.js");
const OuterCircleDataPoint = (props) => {
  const { points, closestPoint, valueAxis = "y", shape } = props;
  const { x, y, width, height } = (0, import_data_point.getOuterCircleSize)(points, valueAxis);
  const strokeColor = (0, import_data_point.getStrokeColor)(points, closestPoint);
  const isGroup = points.length > 1;
  return isGroup ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_RoundRectCanvas.RoundRectCanvas,
    {
      x,
      y,
      width,
      height,
      color: import_colors.default.Background.Base.Default,
      opacity: import_constants2.DEFAULT_HIGHLIGHTED_OPACITY,
      borderRadius: 7,
      borderColor: strokeColor,
      borderWidth: 1
    }
  ) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_SymbolCanvas.SymbolCanvas,
    {
      x: points[0]?.x,
      y: points[0]?.y,
      color: import_colors.default.Background.Base.Default,
      opacity: import_constants2.DEFAULT_HIGHLIGHTED_OPACITY,
      r: import_constants.DATA_POINT_OUTER_RADIUS,
      strokeColor,
      strokeWidth: 1,
      shape
    }
  );
};
const DataPointCanvas = (props) => {
  const { points, valueAxis = "y", shape } = props;
  const paintedPositions = /* @__PURE__ */ new Set();
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    points.length > 0 && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(OuterCircleDataPoint, { ...props }),
    points.map((point, index) => {
      const key = Math.round(point[valueAxis]);
      if (paintedPositions.has(key) || point.id === "nothing") {
        return null;
      }
      paintedPositions.add(key);
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_SymbolCanvas.SymbolCanvas,
        {
          x: point.x,
          y: point.y,
          color: point.color.Default,
          r: import_constants2.DEFAULT_DATA_POINT_RADIUS,
          shape
        },
        index
      );
    })
  ] });
};
