// @dynatrace/strato-components/esm/filters/filter/FilterBar.js
import { jsx, jsxs } from "react/jsx-runtime";
import {
  forwardRef,
  useCallback,
  useEffect,
  useMemo,
  useReducer,
  useState,
  Children
} from "react";
import { IntlProvider } from "react-intl";
import { isElement } from "react-is";
import { getLanguage } from "@dynatrace-sdk/user-preferences";
import { FilterBarContext } from "./contexts/FilterBarContext.js";
import { Item } from "./Item.js";
import { MoreMenu } from "./MoreMenu.js";
import { ResetButton } from "./ResetButton.js";
import { reducer } from "./state/filter-state-reducer.js";
import {
  FilterBarActions
} from "./types/filter-state.types.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { isSlot } from "../../core/utils/is-slot.js";
import { Flex } from "../../layouts/flex/Flex.js";
const COMPONENT_NAME = "FilterBar";
function getItemSlotsAndDefaultPinnedFilters(children, defaultPinnedState) {
  const slots = {
    items: [],
    more: [],
    reset: null
  };
  const defaultPinnedFilters = [];
  for (const child of Children.toArray(children)) {
    if (isSlot(child, Item)) {
      const pinnedState = defaultPinnedState[child.props.name] ?? "pinned";
      switch (pinnedState) {
        case "optional":
          slots.more.push(child);
          break;
        case "pinned-optional":
          defaultPinnedFilters.push(child.props.name);
          slots.more.push(child);
          break;
        default:
          slots.items.push(child);
      }
    } else if (isElement(child) && child.type === ResetButton) {
      slots.reset = child;
    } else if (child !== null) {
      console.warn("The used child is not supported by the FilterBar.");
    }
  }
  return { slots, defaultPinnedFilters };
}
const _FilterBar = /* @__PURE__ */ forwardRef(
  (props, forwardedRef) => {
    const {
      children,
      id,
      onFilterChange,
      onPinnedStateChange,
      defaultPinnedState,
      pinnedState,
      showLabels = true,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps] = useBehavioralTrackingProps(remainingProps);
    const language = getLanguage();
    const [lastAddedItem, setLastAddedItem] = useState(null);
    const [isInitialized, setIsInitialized] = useState(false);
    const callback = useCallback(
      (state2, action) => (
        // eslint-disable-next-line @typescript-eslint/prefer-ts-expect-error, @typescript-eslint/ban-ts-comment
        // @ts-ignore Strict mode transition - needs to be investigated
        reducer(state2, action, !!pinnedState, onPinnedStateChange)
      ),
      // pinnedState is excluded since if it would change the component would switch from controlled to uncontrolled.
      // eslint-disable-next-line react-hooks/exhaustive-deps
      [onFilterChange, onPinnedStateChange]
    );
    const [state, dispatch] = useReducer(callback, {
      filters: void 0,
      itemsPinnedState: defaultPinnedState ?? pinnedState
    });
    useEffect(() => {
      if (pinnedState) {
        dispatch({
          type: FilterBarActions.CHANGEPINNEDSTATE,
          payload: {
            newPinnedState: pinnedState
          }
        });
      }
    }, [pinnedState]);
    useEffect(() => {
      if (state.filters === void 0) {
        return;
      }
      if (!isInitialized) {
        setIsInitialized(true);
      } else {
        onFilterChange(state.filters);
      }
    }, [state.filters]);
    const { slots } = useMemo(
      () => getItemSlotsAndDefaultPinnedFilters(
        children,
        defaultPinnedState ?? pinnedState ?? {}
      ),
      /**
       * Disabled for defaultPinnedState, since this should only be considered when rendered initially.
       */
      // eslint-disable-next-line react-hooks/exhaustive-deps
      [children, pinnedState]
    );
    return /* @__PURE__ */ jsx(
      FilterBarContext.Provider,
      {
        value: {
          state,
          showLabels,
          dispatch,
          lastAddedItem,
          setLastAddedItem,
          hasControlledPinnedState: !!pinnedState
        },
        children: /* @__PURE__ */ jsxs(
          Flex,
          {
            flexWrap: "wrap",
            id,
            ref: forwardedRef,
            "data-testid": dataTestId,
            "data-dtrum-mask": dataDtrumMask,
            "data-dtrum-allow": dataDtrumAllow,
            alignItems: "end",
            className: consumerClassName,
            style: consumerStyle,
            ...behavioralTrackingProps,
            "data-dt-component": COMPONENT_NAME,
            children: [
              slots.items,
              slots.more?.length > 0 && /* @__PURE__ */ jsx(IntlProvider, { locale: language, children: /* @__PURE__ */ jsx(MoreMenu, { children: slots.more }) }),
              slots.reset
            ]
          }
        )
      }
    );
  }
);
_FilterBar.displayName = COMPONENT_NAME;
const FilterBar = Object.assign(_FilterBar, {
  Item,
  ResetButton
});
export {
  FilterBar
};
