// @dynatrace/strato-components/esm/filters/filter/Item.js
import { jsx, jsxs } from "react/jsx-runtime";
import { createElement } from "react";
import {
  useEffect,
  useRef,
  cloneElement
} from "react";
import { useFilterBarContext } from "./hooks/useFilterBarContext.js";
import { FilterBarActions } from "./types/filter-state.types.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useControllableState } from "../../core/hooks/useControllableState.js";
import { isDev } from "../../core/utils/is-node-environment.js";
import { mergeProps as _mergeProps } from "../../core/utils/merge-props.js";
import { FormField } from "../../forms/forms-core/components/form-field/FormField.js";
import { Label } from "../../forms/forms-core/components/label/Label.js";
import { isFormControlWithOverlayRef } from "../../forms/shared-types.js";
const InternalItem = (props) => {
  const {
    name,
    label,
    children: child,
    showLabel: propShowLabel,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps] = useBehavioralTrackingProps(remainingProps);
  const { defaultValue, value, onChange, ...propsWithoutContractProperties } = child.props;
  const [filterValue, setFilterValue] = useControllableState({
    defaultValue,
    value,
    onChange
  });
  const { dispatch, showLabels, lastAddedItem, state } = useFilterBarContext();
  const showLabel = propShowLabel ?? showLabels;
  const isStringLabel = typeof label === "string";
  const filterItemRef = useRef(
    null
  );
  if (!isStringLabel && !showLabel && isDev()) {
    console.warn(
      "FilterBar.Item: the showLabel prop was set to false but the provided label can't be used as aria-label because it is not a string. Please provide a string-based label or set showLabel to true for this FilterBar.Item."
    );
  }
  useEffect(() => {
    dispatch({
      type: FilterBarActions.INITIALIZE,
      payload: {
        name,
        value: filterValue
      }
    });
    setTimeout(() => {
      if (lastAddedItem === name) {
        filterItemRef?.current?.focus?.();
        if (filterItemRef?.current && isFormControlWithOverlayRef(filterItemRef?.current)) {
          filterItemRef?.current?.open();
        }
      }
    });
    return () => {
      dispatch({
        type: FilterBarActions.REMOVE,
        payload: {
          name
        }
      });
    };
  }, [name]);
  useEffect(() => {
    if (state.filters?.[name] && filterValue !== state.filters?.[name].value) {
      dispatch({
        type: FilterBarActions.CHANGE,
        payload: {
          name,
          value: filterValue
        }
      });
    }
  }, [filterValue, state.filters?.[name]]);
  return /* @__PURE__ */ jsxs(
    FormField,
    {
      "data-testid": dataTestId,
      id: propsWithoutContractProperties.id,
      className: consumerClassName,
      style: consumerStyle,
      ...behavioralTrackingProps,
      children: [
        showLabel && /* @__PURE__ */ jsx(Label, { children: label }),
        cloneElement(child, {
          ..._mergeProps(propsWithoutContractProperties, {
            value: filterValue,
            onChange: setFilterValue,
            "aria-label": isStringLabel && !showLabel ? label : void 0,
            ref: filterItemRef
          })
        })
      ]
    }
  );
};
const Item = (props) => {
  return /* @__PURE__ */ createElement(InternalItem, { ...props, key: props.name });
};
Item["displayName"] = "FilterBar.Item";
export {
  Item
};
