// @dynatrace/strato-components/esm/filters/filter/MoreMenu.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import {
  useMemo,
  useRef,
  Children
} from "react";
import { useIntl } from "react-intl";
import { ChevronDownSmallIcon } from "@dynatrace/strato-icons";
import { useFilterBarContext } from "./hooks/useFilterBarContext.js";
import { Item } from "./Item.js";
import { FilterBarActions } from "./types/filter-state.types.js";
import { getTextFromNode } from "./utils/getTextFromNode.js";
import { Button } from "../../buttons/button/Button.js";
import { useId } from "../../core/hooks/useId.js";
import { useFocusRing } from "../../core/styles/useFocusRing.js";
import { isSlot } from "../../core/utils/is-slot.js";
import { Select } from "../../forms/select/Select.js";
import { Flex } from "../../layouts/flex/Flex.js";
const MoreMenu = ({ children }) => {
  const intl = useIntl();
  const label = intl.formatMessage({
    defaultMessage: "Add filter",
    id: "+BUznwQEJTiILiAa",
    description: "Message to display by default for adding more Filters to the FilterBar."
  });
  const { focusClassName, focusProps } = useFocusRing({
    color: "neutral",
    focusWithin: true,
    ignoreModality: true
  });
  const { dispatch, setLastAddedItem, state } = useFilterBarContext(), { itemsPinnedState } = state;
  const selectRef = useRef(null);
  const allPinnableFilters = useMemo(() => {
    const temp = {};
    for (const child of Children.toArray(children)) {
      if (isSlot(child, Item)) {
        const uniqueName = child.props.name;
        temp[uniqueName] = {
          elem: child,
          label: getTextFromNode(child.props.label, intl, uniqueName)
        };
      }
    }
    return temp;
  }, [children, intl]);
  const optionalPinnedItems = useMemo(
    () => Object.entries(itemsPinnedState || {}).reduce((keys, [key, value]) => {
      if (value === "pinned-optional") {
        keys.push(key);
      }
      return keys;
    }, []),
    [itemsPinnedState]
  );
  const handleSelectChange = (newValues) => {
    setLastAddedItem(null);
    const union = /* @__PURE__ */ new Set([
      ...newValues ?? [],
      ...optionalPinnedItems || []
    ]);
    if (!optionalPinnedItems || newValues && optionalPinnedItems.length < newValues.length) {
      const filtersToPin = [...union].filter(
        (key) => !optionalPinnedItems?.includes(key)
      );
      selectRef?.current?.close();
      filtersToPin?.length && setLastAddedItem(filtersToPin[0]?.toString());
      dispatch({
        type: FilterBarActions.PIN,
        payload: {
          keys: filtersToPin
        }
      });
      return;
    }
    if (!newValues) {
      dispatch({
        type: FilterBarActions.UNPIN,
        payload: {
          keys: optionalPinnedItems ?? []
        }
      });
    } else if (optionalPinnedItems && optionalPinnedItems.length > newValues.length) {
      const keys = [...union].filter((key) => !newValues?.includes(key));
      dispatch({
        type: FilterBarActions.UNPIN,
        payload: { keys }
      });
    }
  };
  const labelId = useId();
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    optionalPinnedItems.map(
      (key) => allPinnableFilters[key] && /* @__PURE__ */ jsx(Flex, { gap: 2, alignItems: "flex-end", children: allPinnableFilters[key].elem }, key)
    ),
    /* @__PURE__ */ jsx("div", { style: { maxWidth: "100%" }, ...focusProps, children: /* @__PURE__ */ jsxs(
      Select,
      {
        value: optionalPinnedItems,
        onChange: handleSelectChange,
        id: "user-select",
        multiple: true,
        clearable: true,
        ref: selectRef,
        "data-dt-name": "AddFilter",
        children: [
          /* @__PURE__ */ jsx(Select.CustomTrigger, { children: /* @__PURE__ */ jsxs(Button, { as: "div", id: labelId, className: focusClassName, children: [
            label,
            /* @__PURE__ */ jsx(Button.Suffix, { children: /* @__PURE__ */ jsx(ChevronDownSmallIcon, {}) })
          ] }) }),
          /* @__PURE__ */ jsxs(Select.Content, { width: "fit-content", children: [
            /* @__PURE__ */ jsx(Select.Filter, {}),
            Object.keys(allPinnableFilters).map((key) => /* @__PURE__ */ jsx(Select.Option, { value: key, children: allPinnableFilters[key].label }, key))
          ] })
        ]
      }
    ) })
  ] });
};
MoreMenu.displayName = "FilterBar.MoreMenu";
export {
  MoreMenu
};
