// @dynatrace/strato-components/esm/filters/filter/ResetButton.js
import { jsx } from "react/jsx-runtime";
import { forwardRef, useMemo } from "react";
import { useIntl } from "react-intl";
import { ResetIcon } from "@dynatrace/strato-icons";
import { Button } from "../../buttons/button/Button.js";
import { useId } from "../../core/hooks/useId.js";
import { Tooltip } from "../../overlays/tooltip/Tooltip.js";
const ResetButton = /* @__PURE__ */ forwardRef(
  (props, forwardedRef) => {
    const {
      onClick,
      "aria-labelledby": ariaLabelledByProp,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      ...remainingProps
    } = props;
    const intl = useIntl();
    const resetText = intl.formatMessage({
      defaultMessage: "Reset filters",
      id: "21mG229oVCMPPsYY",
      description: "Shown for the reset button of the FilterBar."
    });
    const tooltipId = useId();
    const ariaLabelledBy = useMemo(
      () => tooltipId || ariaLabelledByProp ? `${tooltipId} ${ariaLabelledByProp ?? ""}` : void 0,
      [ariaLabelledByProp, tooltipId]
    );
    return /* @__PURE__ */ jsx(Tooltip, { id: tooltipId, text: resetText, children: /* @__PURE__ */ jsx(
      Button,
      {
        "data-testid": dataTestId,
        ref: forwardedRef,
        variant: "default",
        onClick,
        "aria-label": resetText,
        "aria-labelledby": ariaLabelledBy,
        className: consumerClassName,
        style: consumerStyle,
        ...remainingProps,
        children: /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(ResetIcon, {}) })
      }
    ) });
  }
);
ResetButton.displayName = "FilterBar.ResetButton";
export {
  ResetButton
};
