// @dynatrace/strato-components/esm/charts/core/components/overlay-tooltip/hooks/useDynamicTooltip.js
import { useRef } from "react";
import { usePopper } from "react-popper";
import { useOverlayStacking } from "../../../../../core/hooks/useOverlayStacking.js";
import { createDOMRect } from "../../../../../core/utils/create-dom-rect.js";
const OPTIONS = {
  strategy: "fixed",
  placement: "right-start",
  modifiers: [
    {
      name: "arrow",
      enabled: false
    },
    {
      name: "offset",
      options: {
        offset: [-10, 10]
      }
    },
    {
      name: "eventListeners",
      enabled: false
    },
    {
      name: "flip",
      options: {
        padding: 10,
        flipVariations: false
      }
    }
  ]
};
const useDynamicTooltip = (args) => {
  const { template, x, y } = args;
  const virtualElement = useRef({
    x: -1,
    y: -1,
    getBoundingClientRect: createDOMRect({ x: -1, y: -1 })
  });
  const overlayStacking = useOverlayStacking(0, "overlay");
  if (x !== virtualElement.current.x || y !== virtualElement.current.y) {
    virtualElement.current = {
      x,
      y,
      getBoundingClientRect: createDOMRect({ x, y })
    };
  }
  const { styles, attributes } = usePopper(
    virtualElement.current,
    template,
    OPTIONS
  );
  return {
    style: {
      ...styles.popper,
      zIndex: `${overlayStacking.style.zIndex}`
    },
    ...attributes.popper
  };
};
export {
  useDynamicTooltip
};
