// @dynatrace-sdk/shared-errors/esm/index.js
/**
 * @license
 * Copyright 2023-2025 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// packages/shared-errors/src/types.ts
var ErrorType = /* @__PURE__ */ ((ErrorType2) => {
  ErrorType2["COMMON"] = "JS Error";
  ErrorType2["HTTP"] = "Http Error";
  return ErrorType2;
})(ErrorType || {});

// packages/shared-errors/src/api-client-error.ts
var ApiClientError = class extends Error {
  constructor(name, message, cause) {
    super(message ?? "ApiClientError: an unexpected error occurred.");
    this.cause = cause;
    this.name = name;
  }
  isApiClientError = true;
  errorType = "JS Error" /* COMMON */;
};
function isApiClientError(e) {
  return e?.isApiClientError === true && e instanceof Error;
}

// packages/shared-errors/src/client-request-error.ts
var ClientRequestError = class extends ApiClientError {
  isClientRequestError = true;
  body;
  response;
  constructor(name, response, body, message, cause) {
    super(name, message, cause);
    this.errorType = "Http Error" /* HTTP */;
    this.body = body;
    this.response = response;
  }
};
function isClientRequestError(e) {
  return e?.isClientRequestError === true && isApiClientError(e);
}

// packages/shared-errors/src/api-gateway-error.ts
var ApiGatewayError = class extends ClientRequestError {
  isApiGatewayError = true;
  code;
  retryAfterSeconds;
  constructor(response, body) {
    super(body.error.details?.errorCode || "ApiGatewayError", response, body, body.error.message);
    this.code = body.error.code;
    this.retryAfterSeconds = body.error.retryAfterSeconds;
  }
};
function isApiGatewayError(e) {
  return e?.isApiGatewayError === true && isClientRequestError(e);
}

// packages/shared-errors/src/invalid-response-error.ts
var InvalidResponseError = class extends ApiClientError {
  isInvalidResponseError = true;
  responseBody;
  expectedType;
  nestedError;
  response;
  constructor(name, nestedError, body, expectedType, message, response) {
    super(
      name,
      message ?? `${name}: Response does not match expected datatype${expectedType ? " " + expectedType : ""}: ${nestedError?.toString() ?? "unable to deserialize"}`
    );
    this.nestedError = nestedError;
    this.responseBody = body;
    this.expectedType = expectedType;
    this.response = response;
  }
};
function isInvalidResponseError(e) {
  return e?.isInvalidResponseError === true && isApiClientError(e);
}
export {
  ApiClientError,
  ApiGatewayError,
  ClientRequestError,
  ErrorType,
  InvalidResponseError,
  isApiClientError,
  isApiGatewayError,
  isClientRequestError,
  isInvalidResponseError
};
