// @dynatrace/strato-components/esm/notifications/notification-settings/hooks/toast-alerts.js
import { jsx, jsxs } from "react/jsx-runtime";
import * as styles from "../../../buttons/notify-button/NotifyButton.sty.js";
import { Text } from "../../../typography/text/Text.js";
import { showToast } from "../../toast/Toast.js";
import { CopyPermissionsButton } from "../components/CopyPermissionsButton.js";
import { notificationMessages } from "../messages.js";
function ToastMessage({ children }) {
  return /* @__PURE__ */ jsx(Text, { className: styles.toast, children });
}
function noWritePermissionToast(intl, notificationEnabled) {
  showToast({
    title: intl.formatMessage(notificationMessages.noPermissionsHeading, {
      notificationEnabled
    }),
    message: /* @__PURE__ */ jsxs(ToastMessage, { children: [
      intl.formatMessage(notificationMessages.copyPermissions),
      "."
    ] }),
    actions: /* @__PURE__ */ jsx(CopyPermissionsButton, { readOrWrite: "write" }),
    position: "bottom-left",
    type: "critical"
  });
}
function notificationLimitReachedToast(intl, notificationType) {
  showToast({
    title: intl.formatMessage(notificationMessages.limitReachedToastHeading),
    message: /* @__PURE__ */ jsx(ToastMessage, { children: intl.formatMessage(notificationMessages.limitReached, {
      notificationType
    }) }),
    position: "bottom-left",
    type: "critical"
  });
}
function notificationGenericErrorToast(intl) {
  showToast({
    title: intl.formatMessage(notificationMessages.genericError),
    message: /* @__PURE__ */ jsx(ToastMessage, { children: intl.formatMessage(notificationMessages.genericErrorMessage) }),
    position: "bottom-left",
    type: "critical"
  });
}
export {
  noWritePermissionToast,
  notificationGenericErrorToast,
  notificationLimitReachedToast
};
