// @dynatrace/strato-components/esm/notifications/notification-settings/hooks/useNotificationSettingsMenuItem.js
import { jsx, jsxs } from "react/jsx-runtime";
import { useMemo, useState } from "react";
import { FormattedMessage } from "react-intl";
import { NotificationIcon } from "@dynatrace/strato-icons";
import { Menu } from "../../../navigation/menu/Menu.js";
import { Modal } from "../../../overlays/modal/Modal.js";
import { isEventNotificationConfig } from "../components/EventNotificationClient.js";
import { isResourceNotificationConfig } from "../components/ResourceNotificationClient.js";
import { notificationMessages } from "../messages.js";
import { NotificationSettings } from "../NotificationSettings.js";
function useNotificationSettingsMenuItem(props) {
  const { type, config, content } = props;
  const [modalOpen, setModalOpen] = useState(false);
  const NotificationMenuItem = useMemo(() => {
    return () => /* @__PURE__ */ jsxs(Menu.Item, { onSelect: () => setModalOpen(true), children: [
      /* @__PURE__ */ jsx(Menu.Prefix, { children: /* @__PURE__ */ jsx(NotificationIcon, {}) }),
      /* @__PURE__ */ jsx(FormattedMessage, { ...notificationMessages.menuItemLabel })
    ] });
  }, []);
  const NotificationSettingsModal = useMemo(() => {
    let notificationSettings = null;
    if (isResourceNotificationConfig(type, config)) {
      notificationSettings = /* @__PURE__ */ jsx(NotificationSettings, { type: "resource", config, children: content });
    } else if (isEventNotificationConfig(type, config)) {
      notificationSettings = /* @__PURE__ */ jsx(NotificationSettings, { type: "event", config, children: content });
    }
    return () => /* @__PURE__ */ jsx(
      Modal,
      {
        show: modalOpen,
        size: "small",
        onDismiss: () => setModalOpen(false),
        title: /* @__PURE__ */ jsx(
          FormattedMessage,
          {
            ...notificationMessages.notificationSettingsModalTitle
          }
        ),
        children: notificationSettings
      }
    );
  }, [config, content, modalOpen, type]);
  return [NotificationMenuItem, NotificationSettingsModal];
}
export {
  useNotificationSettingsMenuItem
};
