// @dynatrace/strato-components/esm/notifications/notification-settings/hooks/useNotificationClient.js
import { useMutation, useQuery } from "@tanstack/react-query";
import { useCallback, useContext } from "react";
import { useIntl } from "react-intl";
import { NotificationQueryClientContext } from "./notification-query-client.js";
import {
  notificationGenericErrorToast,
  notificationLimitReachedToast,
  noWritePermissionToast
} from "./toast-alerts.js";
import { NotifyButtonContext } from "../../../buttons/notify-button/contexts/NotifyButtonContext.js";
function useNotificationClient(notificationType, config, connectors) {
  const intl = useIntl();
  const notificationQueryClient = useContext(NotificationQueryClientContext);
  const { closeOverlay } = useContext(NotifyButtonContext);
  const {
    data: notificationData,
    error: notificationListError,
    isLoading: notificationIsLoading
  } = useQuery(
    {
      initialData: null,
      queryKey: [config, notificationType],
      queryFn: async ({ signal }) => {
        const data = await connectors.list(config, signal);
        const firstResult = data.results.at(0);
        if (firstResult) {
          return firstResult;
        }
        return null;
      }
    },
    notificationQueryClient
  );
  const {
    mutate: createNotification,
    error: notificationCreateError,
    isPending: isCreatePending
  } = useMutation(
    {
      mutationKey: [config, notificationType, "create"],
      mutationFn: async () => {
        return await connectors.create(config);
      },
      throwOnError: false,
      // Handle various error scenarios to prompt user information.
      onError: (error) => {
        if (error.name === "403") {
          noWritePermissionToast(intl, notificationData !== null);
          closeOverlay?.();
        } else if (error.name === "400" && connectors.isErrorEnvelopeError(error) && error.body.error.details?.errorCode?.includes(
          "NotificationLimitReached"
        )) {
          notificationLimitReachedToast(intl, notificationType);
          closeOverlay?.();
        } else {
          notificationGenericErrorToast(intl);
        }
      },
      // Update the query cache to avoid the necessary refetch. The create service
      // already responds with the created notification, so we can use that one in the
      // state.
      onSuccess: (data) => {
        notificationQueryClient.setQueryData([config, notificationType], data);
      }
    },
    notificationQueryClient
  );
  const {
    mutate: deleteNotification,
    error: notificationDeleteError,
    isPending: isDeletePending
  } = useMutation(
    {
      mutationKey: [config, notificationType, "delete"],
      mutationFn: async () => {
        if (notificationData?.id) {
          return connectors.destroy(notificationData.id);
        }
      },
      throwOnError: false,
      onSuccess: (data) => {
        notificationQueryClient.setQueryData([config, notificationType], null);
      },
      onError: (error) => {
        if (error.name === "403") {
          noWritePermissionToast(intl, notificationData !== null);
          closeOverlay?.();
        }
      }
    },
    notificationQueryClient
  );
  const toggleNotifications = useCallback(async () => {
    closeOverlay?.();
    if (notificationData?.id) {
      deleteNotification();
    } else {
      createNotification();
    }
  }, [
    closeOverlay,
    createNotification,
    deleteNotification,
    notificationData?.id
  ]);
  return {
    toggleNotifications,
    enabled: notificationData !== null,
    loading: notificationIsLoading || isCreatePending || isDeletePending,
    hasNotFoundError: notificationListError?.name === "404",
    hasReadPermission: notificationListError?.name !== "403",
    hasWritePermission: notificationCreateError?.name !== "403" && notificationDeleteError?.name !== "403"
  };
}
export {
  useNotificationClient
};
