// @dynatrace/strato-components/esm/charts/histogram/providers/HistogramDataProvider.js
import { jsx } from "react/jsx-runtime";
import { useColorRule } from "../../core/color-rule/useColorRule.js";
import { getColorPalette } from "../../core/utils/get-color-palette.js";
import { HistogramDataContext } from "../contexts/HistogramData.context.js";
import { useHistogramColorPalette } from "../hooks/useHistogramColorPalette.js";
const LEFT_SCALE_INDEX = 0;
function useBuildInitialMetadata(data) {
  const allUnits = /* @__PURE__ */ new Set();
  const colorPalette = useHistogramColorPalette();
  const seriesMap = /* @__PURE__ */ new WeakMap();
  const ruleColor = useColorRule(data, "histogram");
  for (let i = 0; i < data.length; i++) {
    const series = data[i];
    const id = crypto.randomUUID();
    const color = ruleColor.for(series) || (series.color ? { Default: series.color } : getColorPalette(series.name, i, colorPalette));
    allUnits.add(series.unit);
    const metadata = {
      id,
      color,
      scaleIndex: LEFT_SCALE_INDEX
    };
    seriesMap.set(series, metadata);
  }
  return {
    yUnits: [...allUnits],
    seriesMap
  };
}
function safeGet(map, key) {
  const metadata = map.get(key);
  if (!metadata) {
    throw new Error(`Metadata not found for: ${JSON.stringify(key)}`);
  }
  return metadata;
}
const HistogramDataProvider = ({
  children,
  data
}) => {
  const { yUnits, seriesMap } = useBuildInitialMetadata(data);
  return /* @__PURE__ */ jsx(
    HistogramDataContext.Provider,
    {
      value: {
        series: data,
        seriesMap,
        yUnits
      },
      children
    }
  );
};
HistogramDataProvider["displayName"] = "HistogramDataProvider";
export {
  HistogramDataProvider,
  safeGet,
  useBuildInitialMetadata
};
