// @dynatrace/strato-design-tokens/timings-js/index.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var timings_js_exports = {};
__export(timings_js_exports, {
  default: () => timings_js_default
});
module.exports = __toCommonJS(timings_js_exports);
const Instant = {
  /** Value: 0. */
  Default: 0
};
const Offset = {
  /** Value: 0.02. */
  Short: 0.02,
  /** Value: 0.03. */
  Medium: 0.03,
  /** Value: 0.1. */
  Long: 0.1,
  /** Value: 0.05. */
  Default: 0.05
};
const Fast = {
  /** Value: 0.15. */
  Quick: 0.15,
  /** Value: 0.17. */
  Default: 0.17,
  /** Value: 0.2. */
  Moderate: 0.2
};
const Base = {
  /** Value: 0.25. */
  Quick: 0.25,
  /** Value: 0.3. */
  Default: 0.3,
  /** Value: 0.35. */
  Moderate: 0.35
};
const Slow = {
  /** Value: 0.4. */
  Default: 0.4,
  /** Value: 0.6. */
  Moderate: 0.6,
  /** Value: 0.8. */
  Gentle: 0.8
};
const Slowest = {
  /** Value: 1. */
  Default: 1,
  /** Value: 1.2. */
  Moderate: 1.2,
  /** Value: 1.5. */
  Gentle: 1.5
};
var timings_js_default = {
  Instant,
  Offset,
  Fast,
  Base,
  Slow,
  Slowest
};
