// @dynatrace/strato-components/esm/charts/core/components/overlay-tooltip/_to-be-removed/_OverlayTooltipWIPStory.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import {
  cloneElement,
  isValidElement,
  useEffect,
  useRef,
  useState
} from "react";
import { DummyInteractionsOverlay } from "./DummyInteractionsOverlay.js";
import { HeaderTestData } from "./HeaderTestTooltip.js";
import { TooltipData } from "./MyTooltip.js";
import { Button } from "../../../../../buttons/button/Button.js";
import { Select } from "../../../../../forms/select/Select.js";
import { Flex } from "../../../../../layouts/flex/Flex.js";
import { ToastContainer } from "../../../../../notifications/toast/ToastContainer.js";
import { OverlayProvider } from "../contexts/OverlayContext.js";
import { TooltipProvider, useTooltip } from "../contexts/TooltipContext.js";
import { useTooltipReducer } from "../useTooltipReducer.js";
import { TruncationStory } from "./_TruncationStory.js";
const TooltipProgrammaticActions = () => {
  const dispatch = useTooltipReducer();
  const tooltip = useTooltip();
  let snapshot = tooltip.getState();
  const reset = () => {
    snapshot = tooltip.getState();
    dispatch({ type: "RESET_TOOLTIP" })();
  };
  const restore = () => {
    snapshot.pinned && tooltip.setState(snapshot);
  };
  return /* @__PURE__ */ jsxs(Flex, { gap: 4, children: [
    /* @__PURE__ */ jsx(Button, { variant: "emphasized", onClick: reset, children: "Reset Tooltip" }),
    /* @__PURE__ */ jsx(Button, { variant: "emphasized", onClick: restore, children: "Restore tooltip" })
  ] });
};
const createPinAllHandler = (sub) => () => {
  sub.forEach((action) => {
    action();
  });
};
const usePinAllSetup = () => {
  const sub = /* @__PURE__ */ new Set();
  const onClick = createPinAllHandler(sub);
  return { sub, onClick };
};
const chunkArray = (array, size) => {
  const rows = [];
  for (let i = 0; i < array.length; i += size) {
    rows.push(array.slice(i, i + size));
  }
  return rows;
};
const TooltipGrid = (props) => {
  const {
    sub,
    items,
    itemsPerRow = 3,
    width = "350px",
    minHeight = "200px",
    symbolAlignment
  } = props;
  const rows = chunkArray(items, itemsPerRow);
  return /* @__PURE__ */ jsx(Fragment, { children: rows.map((row, rowIndex) => /* @__PURE__ */ jsx(Flex, { flexDirection: "row", children: row.map((item, itemIndex) => {
    const key = rowIndex * itemsPerRow + itemIndex;
    return /* @__PURE__ */ jsxs(Flex, { flexDirection: "column", "data-testid": item.testId, children: [
      item.description,
      /* @__PURE__ */ jsx(
        Content,
        {
          sub,
          width,
          minHeight,
          getData: item.data,
          children: symbolAlignment && isValidElement(item.template) ? cloneElement(
            item.template,
            { symbolAlignment }
          ) : item.template
        }
      )
    ] }, key);
  }) }, rowIndex)) });
};
const TogglePinned = (props) => {
  const dispatch = useTooltipReducer();
  const isPinned = useRef(false);
  useEffect(() => {
    const cb = () => {
      if (isPinned.current) {
        dispatch({
          type: "RESET_TOOLTIP"
        })();
      } else {
        dispatch({
          type: "SET_PINNED",
          payload: {
            pinned: true,
            absoluteX: 10,
            absoluteY: 30,
            data: props.getData ? props.getData() : TooltipData[0].data()
          }
        })();
      }
      isPinned.current = !isPinned.current;
    };
    props.sub.add(cb);
    return () => {
      props.sub.delete(cb);
    };
  }, []);
  return null;
};
const Content = (props) => {
  return /* @__PURE__ */ jsx(OverlayProvider, { children: /* @__PURE__ */ jsxs(TooltipProvider, { children: [
    props.actions && /* @__PURE__ */ jsx(TooltipProgrammaticActions, {}),
    /* @__PURE__ */ jsx(TogglePinned, { sub: props.sub, getData: props.getData }),
    /* @__PURE__ */ jsx(
      "div",
      {
        style: {
          width: props.width || "200px",
          height: "100%",
          maxWidth: "100%",
          maxHeight: "100%",
          resize: "both",
          minHeight: props.minHeight || "50px",
          borderRight: "1px dashed #DDA0DD",
          overflow: "hidden",
          position: "relative"
        },
        children: /* @__PURE__ */ jsx(
          DummyInteractionsOverlay,
          {
            getData: props.getData || TooltipData[0].data,
            children: props.children
          }
        )
      }
    )
  ] }) });
};
const Selector = function(props) {
  return /* @__PURE__ */ jsx(
    Select,
    {
      value: props.value,
      onChange: (newValue) => {
        props.update(newValue);
      },
      children: /* @__PURE__ */ jsx(Select.Content, { children: props.options.map((option) => /* @__PURE__ */ jsx(Select.Option, { value: option, children: `${option}` }, String(option))) })
    }
  );
};
const OverlayTooltipCustomStory = () => {
  const [nTooltips, setNTooltips] = useState(5);
  const { sub, onClick } = usePinAllSetup();
  const Shared = () => {
    return /* @__PURE__ */ jsxs(Fragment, { children: [
      /* @__PURE__ */ jsx(
        Selector,
        {
          options: [5, 50, 100, 150, 300],
          value: nTooltips,
          update: setNTooltips
        }
      ),
      /* @__PURE__ */ jsx(Button, { onClick, children: "Pin all" })
    ] });
  };
  if (nTooltips === 5) {
    return /* @__PURE__ */ jsxs(
      Flex,
      {
        flexWrap: "nowrap",
        gap: 6,
        style: { marginTop: "10px" },
        flexDirection: "column",
        children: [
          /* @__PURE__ */ jsx(Shared, {}),
          Array.from({ length: 5 }).map((_, index) => {
            return /* @__PURE__ */ jsx(
              Content,
              {
                sub,
                actions: true,
                width: "100%",
                minHeight: "400px",
                children: TooltipData[0].template
              },
              index
            );
          })
        ]
      }
    );
  }
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(Shared, {}),
    Array.from({ length: nTooltips / 10 }).map((_, index) => {
      return /* @__PURE__ */ jsx(
        Flex,
        {
          gap: 6,
          style: { marginTop: "10px" },
          flexDirection: "row",
          children: Array.from({ length: 10 }).map((_2, index2) => {
            return /* @__PURE__ */ jsx(Content, { sub, children: TooltipData[0].template }, index2);
          })
        },
        index
      );
    })
  ] });
};
const OverlayTooltipWIPStory = (props) => {
  const { sub, onClick } = usePinAllSetup();
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(ToastContainer, {}),
    /* @__PURE__ */ jsx(Button, { onClick, children: "Pin all" }),
    /* @__PURE__ */ jsx(
      TooltipGrid,
      {
        sub,
        items: TooltipData,
        symbolAlignment: props?.symbolAlignment
      }
    )
  ] });
};
const OverlayTooltipHeaderStory = () => {
  const { sub, onClick } = usePinAllSetup();
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(Button, { onClick, children: "Pin all" }),
    /* @__PURE__ */ jsx(TooltipGrid, { sub, items: HeaderTestData })
  ] });
};
const OverlayTooltipTruncationStory = () => {
  return /* @__PURE__ */ jsx(TruncationStory, {});
};
export {
  OverlayTooltipCustomStory,
  OverlayTooltipHeaderStory,
  OverlayTooltipTruncationStory,
  OverlayTooltipWIPStory,
  TooltipProgrammaticActions
};
