// @dynatrace/strato-components/esm/charts/core/components/overlay-tooltip/_to-be-removed/actions.js
import { jsx, jsxs } from "react/jsx-runtime";
import { DatabaseIcon, MagnifyingGlassIcon } from "@dynatrace/strato-icons";
import { showToast } from "../../../../../notifications/toast/Toast.js";
import { ChartSeriesAction } from "../../series-action/ChartSeriesAction.js";
const TOASTER_LIFESPAN = 800;
const IS_ASYNC = true;
let toasterCounter = 0;
const seriesActions = (data) => {
  return /* @__PURE__ */ jsxs(ChartSeriesAction, { children: [
    /* @__PURE__ */ jsxs(
      ChartSeriesAction.Item,
      {
        onSelect: () => {
          showToast({
            title: `Inspecting ${data.name} ...`,
            type: "info",
            lifespan: 1e3
          });
        },
        children: [
          /* @__PURE__ */ jsx(ChartSeriesAction.ItemIcon, { children: /* @__PURE__ */ jsx(MagnifyingGlassIcon, {}) }),
          "Inspect"
        ]
      },
      "custom-action"
    ),
    /* @__PURE__ */ jsxs(
      ChartSeriesAction.Item,
      {
        disabled: true,
        onSelect: () => void 0,
        children: [
          /* @__PURE__ */ jsx(ChartSeriesAction.ItemIcon, { children: /* @__PURE__ */ jsx(DatabaseIcon, {}) }),
          "A disabled action"
        ]
      },
      "disabled-action"
    ),
    /* @__PURE__ */ jsx(
      ChartSeriesAction.Intent,
      {
        options: {
          responseProperties: ["dt.share"]
        },
        payload: { "dt.query": "describe logs" },
        onResponse: (response) => {
          if (response.status === "success") {
            showToast({
              title: "done...",
              type: "info",
              lifespan: TOASTER_LIFESPAN
            });
          } else {
            showToast({
              title: response.error.message,
              type: "critical",
              lifespan: TOASTER_LIFESPAN
            });
          }
        },
        children: "Send intent with response"
      }
    )
  ] });
};
const asyncSeriesActions = (data) => {
  if (!IS_ASYNC) {
    return seriesActions(data);
  }
  return new Promise((resolve, reject) => {
    showToast({
      title: `Loading series actions ...: ${toasterCounter++}`,
      type: "warning",
      lifespan: TOASTER_LIFESPAN
    });
    setTimeout(() => {
      if (toasterCounter % 5 === 0) {
        showToast({
          title: `Failed to load series actions ...: ${toasterCounter++}`,
          type: "critical",
          lifespan: TOASTER_LIFESPAN
        });
        reject(new Error("Failed to load actions"));
      } else {
        resolve(seriesActions(data));
      }
    }, 500);
  });
};
export {
  asyncSeriesActions
};
