// @dynatrace/strato-components/esm/charts/core/components/overlay-tooltip/_to-be-removed/DummyInteractionsOverlay.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import { isNil } from "lodash-es";
import { ChartOverlaysElevation } from "../../../constants.js";
import { plotLayout } from "../../chart-grid/ChartGrid.sty.js";
import { HorizontalCrosshair } from "../../crosshair/HorizontalCrosshair.js";
import { VerticalCrosshair } from "../../crosshair/VerticalCrosshair.js";
import { interactionsOverlay } from "../../interactions-overlay/InteractionsOverlay.sty.js";
import { useOverlay } from "../contexts/OverlayContext.js";
import { useTooltip } from "../contexts/TooltipContext.js";
import { useSelector, useTooltipSnapshot } from "../selectors.js";
import { useTooltipReducer } from "../useTooltipReducer.js";
import { isOwnEvent } from "../utils.js";
const DummyInteractionsOverlay = (props) => {
  const dispatch = useTooltipReducer();
  const tooltip = useTooltip();
  const overlay = useOverlay();
  return /* @__PURE__ */ jsxs(
    "div",
    {
      role: "none",
      ref: overlay.overlayRef,
      className: interactionsOverlay,
      onPointerMove: (event) => isOwnEvent(event) && dispatch({
        type: "SET_POSITION",
        payload: {
          data: props.getData(),
          absoluteX: Math.round(event.clientX),
          absoluteY: Math.round(event.clientY)
        }
      })(),
      onPointerLeave: dispatch({ type: "SET_OUT_BOUNDS" }),
      onClick: (event) => {
        if (!isOwnEvent(event)) {
          return;
        }
        if (tooltip.getState().pinned) {
          dispatch({
            type: "SET_PINNED",
            payload: {
              pinned: false,
              data: {
                geometry: "none",
                series: void 0,
                header1: void 0,
                subheader: void 0
              }
            }
          })();
        } else {
          dispatch({
            type: "SET_PINNED",
            payload: {
              pinned: true,
              data: props.getData(),
              absoluteX: Math.round(event.clientX),
              absoluteY: Math.round(event.clientY)
            }
          })();
        }
      },
      style: {
        background: "#E6E6FA",
        position: "absolute",
        width: "98%",
        height: "98%",
        top: "1%",
        left: "1%"
      },
      children: [
        /* @__PURE__ */ jsx(Crosshair, {}),
        props.children
      ]
    }
  );
};
const Crosshair = () => {
  const overlay = useOverlay();
  useSelector(
    ({ absoluteX, absoluteY, pinned: pinned2 }) => `${absoluteX}-${absoluteY}-${pinned2}`
  );
  const { relativeX, relativeY, pinned } = useTooltipSnapshot();
  const bounds = overlay.getBounds();
  if (isNil(bounds && relativeX && relativeY)) {
    return null;
  }
  return /* @__PURE__ */ jsxs(
    "svg",
    {
      className: clsx(
        plotLayout,
        {
          overflow: "visible",
          cursor: "crosshair",
          zIndex: ChartOverlaysElevation.crosshair,
          pointerEvents: "none"
        },
        interactionsOverlay
      ),
      children: [
        /* @__PURE__ */ jsx(
          HorizontalCrosshair,
          {
            dashed: !pinned,
            y: relativeY,
            width: bounds?.width,
            height: bounds?.height,
            chartBounds: bounds,
            leftOffset: 0,
            rightOffset: 0
          }
        ),
        /* @__PURE__ */ jsx(
          VerticalCrosshair,
          {
            dashed: !pinned,
            x: relativeX,
            height: bounds?.height,
            width: bounds?.width,
            chartBounds: bounds
          }
        )
      ]
    }
  );
};
export {
  DummyInteractionsOverlay
};
