// @dynatrace/strato-components/esm/charts/core/components/overlay-tooltip/_to-be-removed/HeaderTestTooltip.js
import { jsx } from "react/jsx-runtime";
import { OverlayTooltip } from "../OverlayTooltip.js";
import { useTooltipData } from "./utils.js";
const HeaderTooltip = ({
  label,
  showRangeDuration = false
}) => {
  const { data } = useTooltipData();
  return /* @__PURE__ */ jsx(OverlayTooltip, { children: /* @__PURE__ */ jsx(
    OverlayTooltip.Header,
    {
      header1: data.header1,
      subheader: data.subheader,
      label,
      showRangeDuration1: showRangeDuration
    }
  ) });
};
const DotTooltip = /* @__PURE__ */ jsx(HeaderTooltip, {});
const NoGeometryTooltip = /* @__PURE__ */ jsx(HeaderTooltip, {});
const DurationTooltip = (label) => /* @__PURE__ */ jsx(HeaderTooltip, { label, showRangeDuration: true });
const DualHeaderTooltip = ({
  header2,
  label,
  label2,
  duration2,
  showRangeDuration = false,
  maxLines
}) => {
  const { data } = useTooltipData();
  return /* @__PURE__ */ jsx(OverlayTooltip, { children: /* @__PURE__ */ jsx(
    OverlayTooltip.Header,
    {
      header1: data.header1,
      header2,
      subheader: data.subheader,
      label,
      label2,
      duration2,
      showRangeDuration1: showRangeDuration,
      maxLines
    }
  ) });
};
const baseSeries = [
  {
    name: "Series 1",
    color: "#3DD598",
    value: "100"
  }
];
const dateRange = (from, to) => [
  new Date(from),
  new Date(to)
];
const DATE_RANGES = {
  secondsChanging: dateRange("2023-06-15T14:30:15", "2023-06-15T14:30:45"),
  minutesChanging: dateRange("2023-06-15T14:15:00", "2023-06-15T14:45:00"),
  hoursChanging: dateRange("2023-06-15T09:00:00", "2023-06-15T17:30:00"),
  daysChanging: dateRange("2023-06-05T10:00:00", "2023-06-25T16:30:00"),
  monthsChanging: dateRange("2023-03-15T10:00:00", "2023-09-20T14:30:00"),
  yearsChanging: dateRange("2022-12-15T10:00:00", "2024-03-20T14:30:00"),
  withSeconds: dateRange("2023-06-15T14:30:15", "2023-06-15T14:30:25"),
  crossDay: dateRange("2023-06-15T22:00:00", "2023-06-16T02:30:00"),
  sameDay: dateRange("2023-06-15T08:00:00", "2023-06-15T18:00:00"),
  fullYear: dateRange("2023-01-01T00:00:00", "2023-12-31T23:59:59"),
  janToJun: dateRange("2023-01-01", "2023-06-30"),
  janOnly: dateRange("2023-01-01", "2023-01-31"),
  fullYearSimple: dateRange("2023-01-01", "2023-12-31")
};
const createTestData = (geometry, header1, subheader) => {
  const headerData = Array.isArray(header1) ? header1 : [header1];
  return () => ({
    geometry,
    header1: headerData,
    ...subheader && { subheader },
    series: baseSeries
  });
};
const HeaderTestData = [
  {
    description: /* @__PURE__ */ jsx("h3", { children: "Dot: Single date" }),
    data: createTestData("dot", /* @__PURE__ */ new Date("2023-06-15")),
    template: DotTooltip
  },
  {
    description: /* @__PURE__ */ jsx("h3", { children: "Dot: Date range (from-to)" }),
    data: createTestData("dot", DATE_RANGES.janOnly),
    template: /* @__PURE__ */ jsx(HeaderTooltip, { label: "" })
  },
  {
    description: /* @__PURE__ */ jsx("h3", { children: "Dot: Date range with subheader" }),
    data: createTestData("dot", DATE_RANGES.fullYearSimple, ["Full Year"]),
    template: DotTooltip
  },
  {
    description: /* @__PURE__ */ jsx("h3", { children: "Rect: Single string" }),
    data: createTestData("rect", "Q3"),
    template: /* @__PURE__ */ jsx(HeaderTooltip, { label: "Quarter" })
  },
  {
    description: /* @__PURE__ */ jsx("h3", { children: "Rect: String range" }),
    data: createTestData("rect", ["Start", "End"]),
    template: /* @__PURE__ */ jsx(HeaderTooltip, {})
  },
  {
    description: /* @__PURE__ */ jsx("h3", { children: "Rect: With subheader" }),
    data: createTestData("rect", "Q3", ["2023"]),
    template: /* @__PURE__ */ jsx(HeaderTooltip, {})
  },
  {
    description: /* @__PURE__ */ jsx("h3", { children: "TreeMap: With parent" }),
    data: createTestData("treemap", "United States"),
    template: /* @__PURE__ */ jsx(HeaderTooltip, {})
  },
  {
    description: /* @__PURE__ */ jsx("h3", { children: "Custom label: Single value" }),
    data: createTestData("dot", "Custom Value"),
    template: /* @__PURE__ */ jsx(HeaderTooltip, { label: "My Custom Label" })
  },
  {
    description: /* @__PURE__ */ jsx("h3", { children: "Custom label: Date range" }),
    data: createTestData("dot", DATE_RANGES.janToJun),
    template: DurationTooltip("Time Period")
  },
  {
    description: /* @__PURE__ */ jsx("h3", { children: "Long string value" }),
    data: createTestData(
      "dot",
      "This is a very long header value that might overflow the container"
    ),
    template: DotTooltip
  },
  {
    description: /* @__PURE__ */ jsx("h3", { children: "Numeric value" }),
    data: createTestData("dot", 12345),
    template: DotTooltip
  },
  {
    description: /* @__PURE__ */ jsx("h3", { children: "Numeric range (from-to)" }),
    data: createTestData("dot", [0, 100]),
    template: DotTooltip
  },
  {
    description: /* @__PURE__ */ jsx("h3", { children: "Date Range: Seconds changing (same minute)" }),
    data: createTestData("dot", DATE_RANGES.secondsChanging),
    template: NoGeometryTooltip
  },
  {
    description: /* @__PURE__ */ jsx("h3", { children: "Date Range: Minutes changing (same hour)" }),
    data: createTestData("dot", DATE_RANGES.minutesChanging),
    template: NoGeometryTooltip
  },
  {
    description: /* @__PURE__ */ jsx("h3", { children: "Date Range: Hours changing (same day)" }),
    data: createTestData("dot", DATE_RANGES.hoursChanging),
    template: NoGeometryTooltip
  },
  {
    description: /* @__PURE__ */ jsx("h3", { children: "Date Range: Days changing (same month)" }),
    data: createTestData("dot", DATE_RANGES.daysChanging),
    template: NoGeometryTooltip
  },
  {
    description: /* @__PURE__ */ jsx("h3", { children: "Date Range: Months changing (same year)" }),
    data: createTestData("dot", DATE_RANGES.monthsChanging),
    template: NoGeometryTooltip
  },
  {
    description: /* @__PURE__ */ jsx("h3", { children: "Date Range: Years changing (different years)" }),
    data: createTestData("dot", DATE_RANGES.yearsChanging),
    template: NoGeometryTooltip
  },
  {
    description: /* @__PURE__ */ jsx("h3", { children: "Date Range: With seconds included" }),
    data: createTestData("dot", DATE_RANGES.withSeconds),
    template: NoGeometryTooltip
  },
  {
    description: /* @__PURE__ */ jsx("h3", { children: "Date Range: Cross-day with weekday" }),
    data: createTestData("dot", DATE_RANGES.crossDay),
    template: NoGeometryTooltip
  },
  {
    description: /* @__PURE__ */ jsx("h3", { children: "Date Range: Same day with subheader" }),
    data: createTestData("dot", DATE_RANGES.sameDay, ["Business Hours"]),
    template: NoGeometryTooltip
  },
  {
    description: /* @__PURE__ */ jsx("h3", { children: "Duration: Date range - seconds difference" }),
    data: createTestData("dot", DATE_RANGES.secondsChanging),
    template: DurationTooltip("Time Period")
  },
  {
    description: /* @__PURE__ */ jsx("h3", { children: "Duration: Date range - minutes difference" }),
    data: createTestData("dot", DATE_RANGES.minutesChanging),
    template: DurationTooltip("Meeting Duration")
  },
  {
    description: /* @__PURE__ */ jsx("h3", { children: "Duration: Date range - hours difference" }),
    data: createTestData("dot", DATE_RANGES.hoursChanging),
    template: DurationTooltip("Work Day")
  },
  {
    description: /* @__PURE__ */ jsx("h3", { children: "Duration: Date range - days difference" }),
    data: createTestData("dot", DATE_RANGES.daysChanging),
    template: DurationTooltip("Project Sprint")
  },
  {
    description: /* @__PURE__ */ jsx("h3", { children: "Duration: Numeric range - small difference" }),
    data: createTestData("dot", [1250, 1388]),
    template: DurationTooltip("Score Range")
  },
  {
    description: /* @__PURE__ */ jsx("h3", { children: "Duration: Numeric range - large difference" }),
    data: createTestData("dot", [1e4, 25e4]),
    template: DurationTooltip("Revenue Range")
  },
  {
    description: /* @__PURE__ */ jsx("h3", { children: "Duration: Date range with subheader" }),
    data: createTestData("dot", DATE_RANGES.fullYear, ["Full Year Analysis"]),
    template: DurationTooltip("Annual Period")
  },
  {
    description: /* @__PURE__ */ jsx("h3", { children: "Duration: Cross-month date range" }),
    data: createTestData("dot", DATE_RANGES.monthsChanging),
    template: DurationTooltip("Seasonal Period")
  },
  {
    description: /* @__PURE__ */ jsx("h3", { children: "Duration: Same day range with subheader" }),
    data: createTestData("dot", DATE_RANGES.sameDay, ["Business Hours"]),
    template: DurationTooltip("Daily Operations")
  },
  {
    description: /* @__PURE__ */ jsx("h3", { children: "Dual: Date range (X) + numeric range (Y)" }),
    data: createTestData("dot", DATE_RANGES.janToJun),
    template: /* @__PURE__ */ jsx(DualHeaderTooltip, { header2: [34.5, 96.6] })
  },
  {
    description: /* @__PURE__ */ jsx("h3", { children: "Dual: With axis labels on both lines" }),
    data: createTestData("dot", DATE_RANGES.hoursChanging),
    template: /* @__PURE__ */ jsx(DualHeaderTooltip, { header2: [0, 100], label: "X Axis", label2: "Y Axis" })
  },
  {
    description: /* @__PURE__ */ jsx("h3", { children: "Dual: With range duration on both lines" }),
    data: createTestData("dot", DATE_RANGES.monthsChanging),
    template: /* @__PURE__ */ jsx(DualHeaderTooltip, { header2: [1e4, 25e4], showRangeDuration: true })
  },
  {
    description: /* @__PURE__ */ jsx("h3", { children: "Dual: With subheader \u2014 all three lines" }),
    data: createTestData("dot", DATE_RANGES.fullYear, [
      "18 data points across 2 series"
    ]),
    template: /* @__PURE__ */ jsx(
      DualHeaderTooltip,
      {
        header2: [34.5, 96.6],
        label2: "Y Axis",
        showRangeDuration: true
      }
    )
  },
  {
    description: /* @__PURE__ */ jsx("h3", { children: "Dual: Identical content" }),
    data: createTestData("dot", [34.5, 96.6]),
    template: /* @__PURE__ */ jsx(DualHeaderTooltip, { header2: [34.5, 96.6] })
  },
  {
    description: /* @__PURE__ */ jsx("h3", { children: "Dual: Truncation" }),
    data: createTestData(
      "dot",
      "This is a very long header value that might overflow the container"
    ),
    template: /* @__PURE__ */ jsx(
      DualHeaderTooltip,
      {
        header2: [
          "Beginning of extremely long categorical range",
          "End of extremely long categorical range"
        ],
        label: "X Axis",
        label2: "Y Axis",
        maxLines: 1
      }
    )
  }
];
export {
  HeaderTestData
};
