// @dynatrace/strato-components/esm/charts/core/components/overlay-tooltip/_to-be-removed/MyTooltip.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import {
  CopyIcon,
  DownloadIcon,
  MagnifyingGlassIcon
} from "@dynatrace/strato-icons";
import { asyncSeriesActions } from "./actions.js";
import { useTooltipData } from "./utils.js";
import { showToast } from "../../../../../notifications/toast/Toast.js";
import { OverlayTooltip } from "../OverlayTooltip.js";
import { isTooltipDataEmpty } from "../utils.js";
const XYTooltip = (props) => {
  const data = useTooltipData().data;
  if (isTooltipDataEmpty(data)) {
    return null;
  }
  return /* @__PURE__ */ jsxs(OverlayTooltip, { symbolAlignment: props.symbolAlignment, children: [
    /* @__PURE__ */ jsx(OverlayTooltip.Header, { header1: data.header1, label: props.label }),
    /* @__PURE__ */ jsx(
      OverlayTooltip.Series,
      {
        activeItem: 0,
        variant: props.variant,
        data: data.series,
        seriesItemTemplate: (current) => {
          return /* @__PURE__ */ jsx(
            OverlayTooltip.SeriesItem,
            {
              ...current,
              relativeValueUnit: "%",
              actions: () => asyncSeriesActions(current)
            }
          );
        }
      }
    ),
    /* @__PURE__ */ jsx(
      OverlayTooltip.Action,
      {
        label: "Copy all values",
        icon: /* @__PURE__ */ jsx(CopyIcon, {}),
        onSelect: () => showToast({
          title: "Copy all values",
          type: "info",
          lifespan: 1e3
        })
      }
    )
  ] });
};
const PieTooltip = (props) => {
  const data = useTooltipData().data;
  if (isTooltipDataEmpty(data)) {
    return null;
  }
  return /* @__PURE__ */ jsx(OverlayTooltip, { symbolAlignment: props.symbolAlignment, children: /* @__PURE__ */ jsx(
    OverlayTooltip.SeriesItem,
    {
      ...data.series[0],
      relativeValueUnit: "%",
      actions: () => asyncSeriesActions(data.series[0])
    }
  ) });
};
const MeterBarTooltip = (props) => {
  const data = useTooltipData().data;
  if (isTooltipDataEmpty(data)) {
    return null;
  }
  return /* @__PURE__ */ jsxs(OverlayTooltip, { symbolAlignment: props.symbolAlignment, children: [
    /* @__PURE__ */ jsx(
      OverlayTooltip.SeriesHeader,
      {
        seriesLabel: "App rating",
        valueLabel: "Amount"
      }
    ),
    /* @__PURE__ */ jsx(
      OverlayTooltip.SeriesItem,
      {
        ...data.series[0],
        actions: () => asyncSeriesActions(data.series[0])
      }
    ),
    /* @__PURE__ */ jsx(
      OverlayTooltip.Action,
      {
        label: "Only label action 1",
        onSelect: () => showToast({
          title: "Only label action 1",
          type: "info",
          lifespan: 1e3
        })
      }
    ),
    /* @__PURE__ */ jsx(
      OverlayTooltip.Action,
      {
        label: "Only label action 2",
        onSelect: () => showToast({
          title: "Only label action 2",
          type: "info",
          lifespan: 1e3
        })
      }
    )
  ] });
};
const TreeMapTooltip = (props) => {
  const data = useTooltipData().data;
  if (isTooltipDataEmpty(data)) {
    return null;
  }
  return /* @__PURE__ */ jsxs(OverlayTooltip, { symbolAlignment: props.symbolAlignment, children: [
    /* @__PURE__ */ jsx(OverlayTooltip.Header, { header1: data.header1 }),
    /* @__PURE__ */ jsx(
      OverlayTooltip.SeriesHeader,
      {
        seriesLabel: "Country",
        valueLabel: "Inhabitants"
      }
    ),
    /* @__PURE__ */ jsx(OverlayTooltip.SeriesItem, { ...data.series[0] })
  ] });
};
const TreeMapParentTooltip = (props) => {
  const data = useTooltipData().data;
  if (isTooltipDataEmpty(data)) {
    return null;
  }
  return /* @__PURE__ */ jsxs(OverlayTooltip, { symbolAlignment: props.symbolAlignment, children: [
    /* @__PURE__ */ jsx(OverlayTooltip.Header, { header1: data.header1 }),
    /* @__PURE__ */ jsx(
      OverlayTooltip.SeriesHeader,
      {
        seriesLabel: "Region",
        valueLabel: "Inhabitants"
      }
    ),
    /* @__PURE__ */ jsx(
      OverlayTooltip.SeriesItem,
      {
        ...data.series[0],
        actions: () => asyncSeriesActions(data.series[0])
      }
    ),
    /* @__PURE__ */ jsx(OverlayTooltip.ActionsPlaceholder, { text: "Click to access actions" }),
    /* @__PURE__ */ jsx(
      OverlayTooltip.Action,
      {
        label: "Copy all values",
        icon: /* @__PURE__ */ jsx(CopyIcon, {}),
        onSelect: () => showToast({
          title: "Copy all values",
          type: "info",
          lifespan: 1e3
        })
      }
    ),
    /* @__PURE__ */ jsx(
      OverlayTooltip.Action,
      {
        label: "Download data",
        icon: /* @__PURE__ */ jsx(DownloadIcon, {}),
        onSelect: () => showToast({
          title: "Download data",
          type: "info",
          lifespan: 1e3
        })
      }
    ),
    /* @__PURE__ */ jsx(
      OverlayTooltip.Action,
      {
        label: "View details",
        onSelect: () => showToast({
          title: "View details",
          type: "info",
          lifespan: 1e3
        })
      }
    )
  ] });
};
const HoneyCombTooltip = (props) => {
  const data = useTooltipData().data;
  if (isTooltipDataEmpty(data)) {
    return null;
  }
  return /* @__PURE__ */ jsxs(OverlayTooltip, { symbolAlignment: props.symbolAlignment, children: [
    /* @__PURE__ */ jsx(
      OverlayTooltip.SeriesHeader,
      {
        seriesLabel: "Severity",
        valueLabel: "Incidents"
      }
    ),
    /* @__PURE__ */ jsx(
      OverlayTooltip.SeriesItem,
      {
        ...data.series[0],
        actions: () => asyncSeriesActions(data.series[0])
      }
    )
  ] });
};
const CustomIconTooltip = (props) => {
  const data = useTooltipData().data;
  if (isTooltipDataEmpty(data)) {
    return null;
  }
  return /* @__PURE__ */ jsxs(OverlayTooltip, { symbolAlignment: props.symbolAlignment, children: [
    /* @__PURE__ */ jsx(OverlayTooltip.SeriesHeader, { seriesLabel: "Category", valueLabel: "Count" }),
    /* @__PURE__ */ jsx(
      OverlayTooltip.Series,
      {
        activeItem: 0,
        data: data.series,
        seriesItemTemplate: (current) => {
          return /* @__PURE__ */ jsxs(Fragment, { children: [
            /* @__PURE__ */ jsx(
              OverlayTooltip.SeriesItem,
              {
                icon: /* @__PURE__ */ jsx(CopyIcon, {}),
                name: "Copy actions",
                value: "Copy all values"
              }
            ),
            /* @__PURE__ */ jsx(
              OverlayTooltip.SeriesItem,
              {
                icon: /* @__PURE__ */ jsx(MagnifyingGlassIcon, {}),
                name: "View details",
                value: "View"
              }
            ),
            /* @__PURE__ */ jsx(
              OverlayTooltip.SeriesItem,
              {
                icon: /* @__PURE__ */ jsx(DownloadIcon, {}),
                name: "Download",
                value: "Download data"
              }
            )
          ] });
        }
      }
    )
  ] });
};
const CustomSymbolEmojiTooltip = (props) => {
  const data = useTooltipData().data;
  if (isTooltipDataEmpty(data)) {
    return null;
  }
  return /* @__PURE__ */ jsxs(OverlayTooltip, { symbolAlignment: props.symbolAlignment, children: [
    /* @__PURE__ */ jsx(OverlayTooltip.SeriesHeader, { seriesLabel: "Category", valueLabel: "Count" }),
    /* @__PURE__ */ jsx(
      OverlayTooltip.Series,
      {
        activeItem: 0,
        data: data.series,
        seriesItemTemplate: (current) => {
          return /* @__PURE__ */ jsxs(Fragment, { children: [
            /* @__PURE__ */ jsx(
              OverlayTooltip.SeriesItem,
              {
                symbol: "\u{1F680}",
                name: "Launches",
                value: "42"
              }
            ),
            /* @__PURE__ */ jsx(OverlayTooltip.SeriesItem, { symbol: "\u{1F30D}", name: "Regions", value: "7" }),
            /* @__PURE__ */ jsx(OverlayTooltip.SeriesItem, { symbol: "A", name: "Alpha", value: "100" }),
            /* @__PURE__ */ jsx(OverlayTooltip.SeriesItem, { symbol: "\u2605", name: "Star", value: "200" }),
            /* @__PURE__ */ jsx(OverlayTooltip.SeriesItem, { symbol: "#", name: "Hash", value: "300" })
          ] });
        }
      }
    )
  ] });
};
const CustomShapeIconsTooltip = (props) => {
  const data = useTooltipData().data;
  if (isTooltipDataEmpty(data)) {
    return null;
  }
  return /* @__PURE__ */ jsxs(OverlayTooltip, { symbolAlignment: props.symbolAlignment, children: [
    /* @__PURE__ */ jsx(OverlayTooltip.SeriesHeader, { seriesLabel: "Shape", valueLabel: "Value" }),
    /* @__PURE__ */ jsx(
      OverlayTooltip.Series,
      {
        activeItem: 0,
        data: data.series,
        seriesItemTemplate: (current) => {
          return /* @__PURE__ */ jsxs(Fragment, { children: [
            /* @__PURE__ */ jsx(
              OverlayTooltip.SeriesItem,
              {
                color: "#3DD598",
                shape: "circle",
                name: "Circle (default)",
                value: "100"
              }
            ),
            /* @__PURE__ */ jsx(
              OverlayTooltip.SeriesItem,
              {
                color: "#EB6B52",
                shape: "square",
                name: "Square",
                value: "200"
              }
            ),
            /* @__PURE__ */ jsx(
              OverlayTooltip.SeriesItem,
              {
                color: "#63C45E",
                shape: "triangle",
                name: "Triangle",
                value: "300"
              }
            ),
            /* @__PURE__ */ jsx(
              OverlayTooltip.SeriesItem,
              {
                color: "#EB6B52",
                shape: "bubble",
                name: "Bubble",
                value: "700"
              }
            )
          ] });
        }
      }
    )
  ] });
};
const XYDualAxisTooltip = (props) => {
  const data = useTooltipData().data;
  if (isTooltipDataEmpty(data)) {
    return null;
  }
  return /* @__PURE__ */ jsxs(OverlayTooltip, { symbolAlignment: props.symbolAlignment, children: [
    /* @__PURE__ */ jsx(
      OverlayTooltip.Header,
      {
        header1: data.header1,
        header2: [/* @__PURE__ */ new Date("2026-01-19 10:30"), /* @__PURE__ */ new Date("2025-02-10 10:55")],
        subheader: data.subheader,
        showRangeDuration1: true
      }
    ),
    /* @__PURE__ */ jsx(
      OverlayTooltip.Series,
      {
        activeItem: 0,
        variant: "single",
        data: data.series,
        seriesItemTemplate: (current) => /* @__PURE__ */ jsx(OverlayTooltip.SeriesItem, { ...current })
      }
    )
  ] });
};
const DotTooltipData = {
  geometry: "line",
  header1: ["Q3"],
  series: [
    {
      name: "R&D",
      color: "#3DD598",
      value: "505K",
      valueUnit: "$"
    },
    {
      name: "Support",
      color: "#EB6B52",
      value: "350K",
      valueUnit: "$",
      relativeValue: void 0,
      relativeValueUnit: void 0
    },
    {
      name: "Sales",
      color: "#63C45E",
      value: "320K",
      valueUnit: "$",
      relativeValue: void 0,
      relativeValueUnit: void 0
    },
    {
      name: "Marketing",
      color: "#D971C0",
      value: "305K",
      valueUnit: "$",
      relativeValue: void 0,
      relativeValueUnit: void 0
    },
    {
      name: "HR",
      color: "#D9B671",
      value: "205K",
      valueUnit: "$",
      relativeValue: void 0,
      relativeValueUnit: void 0
    },
    {
      name: "IT",
      color: "#3DD598",
      value: "105K",
      valueUnit: "$"
    },
    {
      name: "Finance",
      color: "#EB6B52",
      value: "95K",
      valueUnit: "$"
    },
    {
      name: "Admin",
      color: "#63C45E",
      value: "80K",
      valueUnit: "$"
    },
    {
      name: "Management",
      color: "#D971C0",
      value: "75K",
      valueUnit: "$"
    },
    {
      name: "Legal",
      color: "#D9B671",
      value: "65K",
      valueUnit: "$"
    },
    {
      name: "Operations",
      color: "#3DD598",
      value: "55K",
      valueUnit: "$"
    },
    {
      name: "Other",
      color: "#EB6B52",
      value: "50K",
      valueUnit: "$"
    },
    {
      name: "Logistics",
      color: "#63C45E",
      value: "45K",
      valueUnit: "$"
    },
    {
      name: "Procurement",
      color: "#D971C0",
      value: "30K",
      valueUnit: "$"
    }
  ]
};
const rectTooltipData = {
  geometry: "rect",
  header1: [new Date(2023, 3, 1, 7, 30), new Date(2023, 7, 1, 7, 35)],
  series: [
    {
      name: "R&D",
      color: "#3DD598",
      value: "505K $",
      relativeValue: "25"
    }
  ]
};
const rectTooltipNumericData = {
  geometry: "rect",
  header1: [1325, 1388],
  series: [
    {
      name: "HR",
      color: "#D9B671",
      value: "205K $",
      relativeValue: void 0
    }
  ]
};
const headerStyle = { margin: 0, marginBlock: "unset" };
const TooltipData = [
  {
    testId: "dot-with-single-series",
    description: /* @__PURE__ */ jsxs(Fragment, { children: [
      /* @__PURE__ */ jsx("h3", { style: headerStyle, children: "Dot tooltip with single series" }),
      /* @__PURE__ */ jsx("span", { children: "1 Global Action " })
    ] }),
    data: () => DotTooltipData,
    template: /* @__PURE__ */ jsx(XYTooltip, { label: "Quarter", variant: "single" })
  },
  {
    testId: "dot-with-timeframe",
    description: /* @__PURE__ */ jsxs(Fragment, { children: [
      /* @__PURE__ */ jsx("h3", { style: headerStyle, children: "Dot tooltip with timeframe" }),
      /* @__PURE__ */ jsx("span", { children: "1 Global Action " })
    ] }),
    data: () => ({
      ...DotTooltipData,
      header1: [new Date(2023, 0, 1, 7, 30), new Date(2023, 0, 1, 7, 35)]
    }),
    template: /* @__PURE__ */ jsx(XYTooltip, { variant: "single", label: "" })
  },
  {
    testId: "rect-with-single-series",
    description: /* @__PURE__ */ jsxs(Fragment, { children: [
      /* @__PURE__ */ jsx("h3", { style: headerStyle, children: "Rect tooltip with single series" }),
      /* @__PURE__ */ jsx("span", { children: "1 Global Action " })
    ] }),
    data: () => ({
      geometry: "rect",
      header1: [new Date(2023, 0, 1, 7, 30), new Date(2023, 0, 1, 7, 35)],
      series: [
        {
          name: "WebServiceA-CW1-1",
          color: "#3DD598",
          value: 35,
          valueUnit: "ms"
        }
      ]
    }),
    template: /* @__PURE__ */ jsx(XYTooltip, { variant: "single" })
  },
  {
    testId: "dot-multiple-series",
    description: /* @__PURE__ */ jsxs(Fragment, { children: [
      /* @__PURE__ */ jsx("h3", { style: headerStyle, children: "Dot tooltip with multiple series" }),
      /* @__PURE__ */ jsx("span", { children: "1 Global Action " })
    ] }),
    data: () => DotTooltipData,
    template: /* @__PURE__ */ jsx(XYTooltip, { variant: "shared" })
  },
  {
    testId: "rect-with-timeframe",
    description: /* @__PURE__ */ jsxs(Fragment, { children: [
      /* @__PURE__ */ jsx("h3", { style: headerStyle, children: "Rect tooltip with timeframe" }),
      /* @__PURE__ */ jsx("span", { children: "1 Global Action " })
    ] }),
    data: () => rectTooltipData,
    template: /* @__PURE__ */ jsx(XYTooltip, { variant: "single" })
  },
  {
    testId: "rect-with-numeric-range",
    description: /* @__PURE__ */ jsxs(Fragment, { children: [
      /* @__PURE__ */ jsx("h3", { style: headerStyle, children: "Rect tooltip with numeric range" }),
      /* @__PURE__ */ jsx("span", { children: "1 Global Action " })
    ] }),
    data: () => rectTooltipNumericData,
    template: /* @__PURE__ */ jsx(XYTooltip, { variant: "single", label: "" })
  },
  {
    testId: "pie",
    description: /* @__PURE__ */ jsx("h3", { children: "Pie tooltip" }),
    data: () => ({
      geometry: "pie",
      series: [
        {
          name: "Latam",
          color: "#3DD598",
          value: "35K",
          valueUnit: "$",
          relativeValue: "25"
        }
      ]
    }),
    template: /* @__PURE__ */ jsx(PieTooltip, {})
  },
  {
    testId: "meter-bar",
    description: /* @__PURE__ */ jsxs(Fragment, { children: [
      /* @__PURE__ */ jsx("h3", { style: headerStyle, children: "MeterBar tooltip" }),
      /* @__PURE__ */ jsx("span", { children: "2 Global Actions (only label)" })
    ] }),
    data: () => ({
      geometry: "meterbar",
      series: [
        {
          name: "Good",
          color: "#3DD598",
          value: "85"
        }
      ]
    }),
    template: /* @__PURE__ */ jsx(MeterBarTooltip, {})
  },
  {
    testId: "treeMap-with-parent",
    description: /* @__PURE__ */ jsxs(Fragment, { children: [
      /* @__PURE__ */ jsx("h3", { style: headerStyle, children: "TreeMap tooltip with parent" }),
      /* @__PURE__ */ jsx("span", { children: "3 Global Actions" })
    ] }),
    data: () => ({
      geometry: "treemap",
      header1: ["United States"],
      series: [
        {
          name: "New York City",
          color: "#3DD598",
          value: "18.8M"
        }
      ]
    }),
    template: /* @__PURE__ */ jsx(TreeMapParentTooltip, {})
  },
  {
    testId: "treeMap-without-parent",
    description: /* @__PURE__ */ jsx("h3", { children: "TreeMap tooltip without parent" }),
    data: () => ({
      geometry: "treemap",
      header1: [],
      series: [
        {
          name: "Brazil",
          color: "#3DD598",
          value: "35.2M"
        }
      ]
    }),
    template: /* @__PURE__ */ jsx(TreeMapTooltip, {})
  },
  {
    testId: "treeMap-without-parent-no-color",
    description: /* @__PURE__ */ jsx("h3", { children: "TreeMap tooltip without parent and no color" }),
    data: () => ({
      geometry: "treemap",
      header1: [],
      series: [
        {
          name: "Brazil",
          value: "35.2M"
        }
      ]
    }),
    template: /* @__PURE__ */ jsx(TreeMapTooltip, {})
  },
  {
    testId: "honeyComb",
    description: /* @__PURE__ */ jsx("h3", { children: "HoneyComb tooltip" }),
    data: () => ({
      geometry: "honeycomb",
      header1: [],
      series: [
        {
          name: "Emergency",
          color: "#3DD598",
          value: "4.1K"
        }
      ]
    }),
    template: /* @__PURE__ */ jsx(HoneyCombTooltip, {})
  },
  {
    description: /* @__PURE__ */ jsx("h3", { children: "Custom symbol with Icons" }),
    testId: "custom-symbol-with-icons",
    data: () => ({
      geometry: "dot",
      header1: [],
      series: [
        {
          name: "Launches",
          color: "",
          value: "42"
        }
      ]
    }),
    template: /* @__PURE__ */ jsx(CustomIconTooltip, {})
  },
  {
    testId: "custom-symbol-with-emojis-and-characters",
    description: /* @__PURE__ */ jsx("h3", { children: "Custom symbol emojis & characters" }),
    data: () => ({
      geometry: "dot",
      header1: [],
      series: [
        {
          name: "Launches",
          color: "",
          value: "42"
        }
      ]
    }),
    template: /* @__PURE__ */ jsx(CustomSymbolEmojiTooltip, {})
  },
  {
    testId: "shape-icons-circle-square-triangle-etc",
    description: /* @__PURE__ */ jsx("h3", { children: "Shape icons (circle, square, triangle, etc.)" }),
    data: () => ({
      geometry: "dot",
      header1: [],
      series: [
        {
          name: "Circle",
          color: "#3DD598",
          value: "100"
        }
      ]
    }),
    template: /* @__PURE__ */ jsx(CustomShapeIconsTooltip, {})
  },
  {
    testId: "xy-dual-axis-header",
    description: /* @__PURE__ */ jsxs(Fragment, { children: [
      /* @__PURE__ */ jsx("h3", { style: headerStyle, children: "XY tooltip with dual axis header" }),
      /* @__PURE__ */ jsx("span", { children: "Both X and Y axis ranges with duration" })
    ] }),
    data: () => ({
      geometry: "rect",
      header1: [/* @__PURE__ */ new Date("2023-01-19"), /* @__PURE__ */ new Date("2023-02-10")],
      subheader: ["18 data points across 2 series"],
      series: [
        {
          name: "WebServiceA",
          color: "#3DD598",
          value: "65.0",
          valueUnit: "ms"
        }
      ]
    }),
    template: /* @__PURE__ */ jsx(XYDualAxisTooltip, {})
  }
];
export {
  DotTooltipData,
  TooltipData,
  XYDualAxisTooltip
};
