// @dynatrace/strato-components/charts/core/components/annotations/components/AnnotationsRenderer.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var AnnotationsRenderer_exports = {};
__export(AnnotationsRenderer_exports, {
  AnnotationsRenderer: () => AnnotationsRenderer
});
module.exports = __toCommonJS(AnnotationsRenderer_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_OverlayContext = require("../../overlay-tooltip/contexts/OverlayContext.js");
var import_AnnotationsAtom = require("../AnnotationsAtom.js");
var import_selectors = require("../hooks/store/selectors.js");
var import_useGroupedAnnotationsData = require("../hooks/useGroupedAnnotationsData.js");
var import_useMarkerEventHandlers = require("../hooks/useMarkerEventHandlers.js");
var import_MarkerAtom = require("../slots/MarkerAtom.js");
var import_TrackAtom = require("../slots/TrackAtom.js");
var import_calculate_marker_dimension = require("../utils/calculate-marker-dimension.js");
var import_define_marker_brightness_state = require("../utils/define-marker-brightness-state.js");
var import_define_marker_color = require("../utils/define-marker-color.js");
var import_define_marker_content = require("../utils/define-marker-content.js");
const AnnotationsRenderer = (props) => {
  const {
    scale,
    width,
    labelsOffset,
    showLabels,
    tracks,
    textOverflow,
    truncateMode
  } = props;
  const overlay = (0, import_OverlayContext.useOverlay)();
  const groupedData = (0, import_useGroupedAnnotationsData.useGroupedAnnotationsData)(tracks, scale.domain());
  const selection = (0, import_selectors.useAnnotationsSelectionState)();
  const { handleMouseEnter, handleMouseMove, handleMouseLeave, handleClick } = (0, import_useMarkerEventHandlers.useMarkerEventHandlers)(overlay.overlayRef);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_AnnotationsAtom.AnnotationsAtom,
    {
      width,
      labelsWidth: labelsOffset,
      showLabels,
      children: groupedData.map((track, trackNumber) => {
        const currentTrack = Array.from(track.markers.entries());
        return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_TrackAtom.TrackAtom,
          {
            label: track.label,
            hidden: track.hidden,
            children: currentTrack.map(([groupId, annotationsGroup]) => {
              return [...annotationsGroup].sort((a, b) => a.priority - b.priority).map(({ key: markerId, markers }) => {
                const [marker] = markers;
                const color = (0, import_define_marker_color.defineMarkerColor)(markers);
                const state = (0, import_define_marker_brightness_state.defineMarkerBrightnessState)(
                  markerId,
                  selection
                );
                const { x, width: width2 } = (0, import_calculate_marker_dimension.calculateMarkerDimension)(marker, scale);
                const markerContent = (0, import_define_marker_content.defineMarkerContent)(markers);
                return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  import_MarkerAtom.MarkerAtom,
                  {
                    x,
                    width: width2,
                    color,
                    onMouseEnter: handleMouseEnter(
                      markerId,
                      groupId,
                      annotationsGroup
                    ),
                    onMouseMove: handleMouseMove,
                    onMouseLeave: handleMouseLeave,
                    onClick: handleClick(groupId, annotationsGroup),
                    content: markerContent,
                    interactionState: state,
                    textOverflow,
                    truncateMode
                  },
                  markerId
                );
              });
            })
          },
          trackNumber
        );
      })
    }
  );
};
AnnotationsRenderer.displayName = "AnnotationsContent";
