// @dynatrace/strato-components/forms/text-area/TextArea.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TextArea_exports = {};
__export(TextArea_exports, {
  TextArea: () => TextArea
});
module.exports = __toCommonJS(TextArea_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var styles = __toESM(require("./TextArea.sty.js"));
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useControllableState = require("../../core/hooks/useControllableState.js");
var import_useFormControlState = require("../../core/hooks/useFormControlState.js");
var import_useFocusRing = require("../../core/styles/useFocusRing.js");
var import_merge_props = require("../../core/utils/merge-props.js");
var import_Flex = require("../../layouts/flex/Flex.js");
var import_FormFieldMessageContextProvider = require("../forms-core/components/form-field/FormFieldMessageContextProvider.js");
var import_FormFieldMessageBaseItem = require("../forms-core/components/form-field-messages/FormFieldMessageBaseItem.js");
var import_useFormFieldMessagesContext = require("../forms-core/hooks/useFormFieldMessagesContext.js");
var import_useFormFieldValidation = require("../forms-core/hooks/useFormFieldValidation.js");
var import_useSyncedFormControlProps = require("../forms-core/hooks/useSyncedFormControlProps.js");
var import_formfield_messages_state_reducer = require("../forms-core/state/formfield-messages-state-reducer.js");
var import_useInputHint = require("../hooks/useInputHint.js");
const COMPONENT_NAME = "TextArea";
const TextAreaRoot = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const intl = (0, import_react_intl.useIntl)();
  const defaultPlaceholder = intl.formatMessage({
    defaultMessage: "Add text",
    id: "E7kKT3Td1FKB37wr",
    description: "Default placeholder for the text area input."
  });
  const {
    onChange,
    onBlur,
    minLength,
    maxLength,
    id: propId,
    disabled: propDisabled,
    required: propRequired,
    readOnly: propReadOnly,
    resize = "both",
    controlState,
    value: propValue,
    defaultValue,
    width = "full",
    placeholder = defaultPlaceholder,
    "aria-describedby": ariaDescribedby,
    "aria-disabled": ariaDisabledProp,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    className: consumerClassName,
    style: consumerStyle,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const { disabled, required, id, ariaDisabled, readOnly } = (0, import_useSyncedFormControlProps.useSyncedFormControlProps)({
    propDisabled,
    propRequired,
    propId,
    propAriaDisabled: ariaDisabledProp,
    propReadOnly
  });
  const disabledOrAriaDisabled = disabled || ariaDisabled || false;
  const { state: messagesContextState, dispatch: messagesContextDispatch } = (0, import_useFormFieldMessagesContext.useFormFieldMessagesContext)();
  const containerRef = (0, import_react.useRef)(null);
  const textareaRef = (0, import_react.useRef)(null);
  (0, import_react.useImperativeHandle)(forwardedRef, () => {
    return {
      element: containerRef.current,
      inputRef: textareaRef.current,
      focus() {
        textareaRef.current?.focus();
      },
      validate() {
        textareaRef.current?.checkValidity();
        messagesContextDispatch?.({
          type: import_formfield_messages_state_reducer.FormFieldMessagesAction.SETISTOUCHED,
          payload: { isTouched: true }
        });
      }
    };
  });
  const [controlStateHasError, hint] = (0, import_useFormControlState.useFormControlState)(controlState);
  const hasError = controlState ? controlStateHasError : messagesContextState.showError;
  const [value, setValue] = (0, import_useControllableState.useControllableState)({
    defaultValue,
    value: propValue
  });
  const {
    onChange: validationOnChange,
    onBlur: validationOnBlur,
    onInvalid: validationOnInvalid,
    ariaProps
  } = (0, import_useFormFieldValidation.useFormFieldValidation)(textareaRef, { required, minLength, maxLength });
  const handleChange = (event) => {
    const newValue = event.target?.value;
    setValue(newValue);
    onChange?.(newValue, event);
    !controlState && validationOnChange();
  };
  const { inputProps, hintProps } = (0, import_useInputHint.useInputHint)({
    ariaDescribedby,
    hint,
    hasError
  });
  const {
    focusProps: { onFocus, onBlur: focusRingOnBlur },
    focusClassName
  } = (0, import_useFocusRing.useFocusRing)({
    color: hasError ? "critical" : "neutral",
    ignoreModality: true,
    disabled
  });
  const styleVariants = {
    error: hasError,
    disabled: disabledOrAriaDisabled,
    readOnly
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_Flex.Flex,
    {
      gap: 4,
      flexDirection: "column",
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ref: containerRef,
      className: consumerClassName,
      style: consumerStyle,
      ...behavioralTrackingProps,
      "data-dt-component": COMPONENT_NAME,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
          "div",
          {
            className: styles.container({
              contentWidth: width === "content",
              resize
            }),
            style: { width },
            children: [
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                "textarea",
                {
                  ref: textareaRef,
                  id,
                  minLength,
                  maxLength,
                  className: (0, import_clsx.default)(styles.textArea(styleVariants), focusClassName),
                  disabled,
                  required,
                  readOnly,
                  value,
                  "aria-required": required,
                  "aria-disabled": ariaDisabled,
                  placeholder,
                  onChange: handleChange,
                  onInvalid: (e) => !controlState && validationOnInvalid(e),
                  onFocus,
                  onBlur: (e) => {
                    !controlState && validationOnBlur(e);
                    focusRingOnBlur(e);
                    onBlur?.(e);
                  },
                  ...(0, import_merge_props.mergeProps)(restProps, controlState ? inputProps : ariaProps)
                }
              ),
              resize !== "none" && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ResizeIcon, { className: styles.resizeIcon(styleVariants) })
            ]
          }
        ),
        hint && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_FormFieldMessageBaseItem.InternalBaseItem,
          {
            variant: hasError ? "error" : "hint",
            disabled: disabled || ariaDisabled,
            ...hintProps,
            children: hint
          }
        )
      ]
    }
  );
});
const TextArea = (0, import_FormFieldMessageContextProvider.withFormFieldMessageContextProvider)(TextAreaRoot);
TextArea.displayName = COMPONENT_NAME;
