// @dynatrace/strato-components/esm/charts/categorical-bar/utils/categorical-layout.js
import { isStacked } from "./layout.js";
import { safeGet } from "../../core/utils/safe-get.js";
function getTopValueCoord(dimension, valueScale, dimensionMap) {
  const { value, cumulativeValue } = safeGet(dimensionMap, dimension);
  return valueScale((cumulativeValue ?? 0) + value) ?? 0;
}
function getBarSize(value, valueScale) {
  const zeroCoordinate = valueScale(0) ?? 0;
  const scaledValue = valueScale(value) ?? 0;
  if (value >= 0) {
    return zeroCoordinate - scaledValue;
  }
  return scaledValue - zeroCoordinate;
}
function getValueCoord(dimension, valueScale, dimensionMap) {
  const { value, cumulativeValue } = safeGet(dimensionMap, dimension);
  return valueScale(
    value >= 0 ? (cumulativeValue ?? 0) + value : cumulativeValue ?? 0
  ) ?? 0;
}
function getBarCoordinates(groupMode, categoryScale, valueScale, dimensionsScale, isSingleDimension, dimensionMap, categoryMap) {
  if (isStacked(groupMode) || isSingleDimension) {
    return {
      categoryCoord: (dimension, category) => categoryScale(safeGet(categoryMap, category).id) ?? 0,
      valueCoord: (dimension) => getValueCoord(dimension, valueScale, dimensionMap),
      barThickness: () => categoryScale.bandwidth(),
      barSize: (dimension) => getBarSize(safeGet(dimensionMap, dimension).value, valueScale)
    };
  }
  return {
    categoryCoord: (dimension, category) => {
      const categoryCoord = categoryScale(safeGet(categoryMap, category).id) ?? 0;
      const dimensionCoord = dimensionsScale(safeGet(dimensionMap, dimension).name) ?? 0;
      return categoryCoord + dimensionCoord;
    },
    valueCoord: (dimension) => getValueCoord(dimension, valueScale, dimensionMap),
    barThickness: () => dimensionsScale.bandwidth(),
    barSize: (dimension) => getBarSize(safeGet(dimensionMap, dimension).value, valueScale)
  };
}
export {
  getBarCoordinates,
  getBarSize,
  getTopValueCoord
};
