// @dynatrace/strato-components/esm/charts/categorical-bar/utils/categorical-config.js
import { merge } from "lodash-es";
import { isValidGroupMode } from "./is-valid-group-mode.js";
import { isValidTooltipVariant } from "./is-valid-tooltip-variant.js";
import { isValidColorRule } from "../../core/slots/ColorRule.js";
import {
  isCustomColorPalette
} from "../../core/types/color-palette.js";
import { isValidCSSColor } from "../../core/utils/getValidColor.js";
import { isValidColorPalette } from "../../core/utils/is-valid-color-palette.js";
import { isValidLegendPosition } from "../../core/utils/is-valid-legend-position.js";
import { isValidValueRepresentation } from "../../core/utils/is-valid-value-representation.js";
import { DEFAULT_SCALE_FALLBACK_COLOR } from "../constants.js";
import { defaultConfig } from "../types/defaults.js";
import { isValidLayout } from "../utils/is-valid-layout.js";
function getColorPaletteConfiguration(contextColorPalette, propsColorPalette) {
  const colorPalette = propsColorPalette ?? contextColorPalette ?? defaultConfig.colorPalette;
  return isCustomColorPalette(colorPalette) || isValidColorPalette(colorPalette) ? colorPalette : defaultConfig.colorPalette;
}
function getColorPaletteMode(propsColorPaletteMode, contextColorPaletteMode) {
  const colorPaletteMode = propsColorPaletteMode ?? contextColorPaletteMode ?? defaultConfig.colorPaletteMode;
  return colorPaletteMode;
}
function getLayoutConfiguration(contextLayout, propsLayout) {
  const layout = propsLayout ?? contextLayout ?? defaultConfig.layout;
  return isValidLayout(layout) ? layout : defaultConfig.layout;
}
function getGroupModeConfiguration(contextGroupMode, propsGroupMode) {
  const groupMode = propsGroupMode ?? contextGroupMode ?? defaultConfig.groupMode;
  return isValidGroupMode(groupMode) ? groupMode : defaultConfig.groupMode;
}
function getValueRepresentationConfiguration(contextValueRepresentation, propsValueRepresentation) {
  const valueRepresentation = propsValueRepresentation ?? contextValueRepresentation ?? defaultConfig.valueRepresentation;
  return isValidValueRepresentation(valueRepresentation) ? valueRepresentation : defaultConfig.valueRepresentation;
}
function getThresholdsConfig(contextThresholds = [], slotsThresholds) {
  return contextThresholds.concat(slotsThresholds).map((threshold) => ({
    ...threshold,
    id: crypto.randomUUID(),
    color: isValidCSSColor(threshold.color) ? threshold.color : DEFAULT_SCALE_FALLBACK_COLOR
  }));
}
function buildCategoricalBarChartConfig(props, contextConfig, slots) {
  const colorPalette = getColorPaletteConfiguration(
    contextConfig.colorPalette,
    props.colorPalette
  );
  const colorPaletteMode = getColorPaletteMode(
    props.colorPaletteMode,
    contextConfig.colorPaletteMode
  );
  const layout = getLayoutConfiguration(contextConfig.layout, props.layout);
  const groupMode = getGroupModeConfiguration(
    contextConfig.groupMode,
    props.groupMode
  );
  const valueRepresentation = getValueRepresentationConfiguration(
    contextConfig.valueRepresentation,
    props.valueRepresentation
  );
  const categoryAxis = merge(
    {},
    defaultConfig.categoryAxis,
    contextConfig.categoryAxis,
    slots.categoryAxis
  );
  categoryAxis.tickLabelLayout = isValidLayout(categoryAxis.tickLabelLayout) ? categoryAxis.tickLabelLayout : defaultConfig.categoryAxis.tickLabelLayout;
  const valueAxis = merge(
    {},
    defaultConfig.valueAxis,
    contextConfig.valueAxis,
    slots.valueAxis
  );
  const legend = merge(
    {},
    defaultConfig.legend,
    contextConfig.legend,
    slots.legend
  );
  legend.position = isValidLegendPosition(legend.position) ? legend.position : defaultConfig.legend.position;
  legend.hidden = Boolean(legend.hidden);
  const tooltip = merge(
    {},
    defaultConfig.tooltip,
    contextConfig.tooltip,
    slots.tooltip
  );
  tooltip.variant = isValidTooltipVariant(tooltip.variant) ? tooltip.variant : defaultConfig.tooltip.variant;
  const toolbar = merge(
    {},
    defaultConfig.toolbar,
    contextConfig.toolbar,
    slots.toolbar
  );
  const truncationMode = props.truncationMode ?? contextConfig.truncationMode ?? defaultConfig.truncationMode;
  const configThresholds = contextConfig.thresholdIndicators || [];
  const thresholdIndicators = getThresholdsConfig(
    configThresholds,
    slots.thresholds
  );
  const colorRules = (contextConfig.colorRules ?? []).filter(isValidColorRule).concat(slots.colorRules ?? []);
  return {
    colorPalette,
    colorPaletteMode,
    layout,
    groupMode,
    valueRepresentation,
    valueAxis,
    categoryAxis,
    legend,
    tooltip,
    toolbar,
    truncationMode,
    thresholdIndicators,
    colorRules,
    intents: slots.intents
  };
}
function deserializeConfig(serializedConfig) {
  try {
    return merge({}, defaultConfig, JSON.parse(serializedConfig));
  } catch {
    return defaultConfig;
  }
}
export {
  buildCategoricalBarChartConfig,
  deserializeConfig,
  getColorPaletteConfiguration,
  getColorPaletteMode,
  getGroupModeConfiguration,
  getLayoutConfiguration,
  getThresholdsConfig,
  getValueRepresentationConfiguration
};
