// @dynatrace/strato-components/esm/charts/categorical-bar/utils/generate-categorical-bar-data-helpers.js
import { isNil, isNull } from "lodash-es";
import {
  isCustomColorPalette
} from "../../core/types/color-palette.js";
import { getColorPalette } from "../../core/utils/get-color-palette.js";
import { isValidColorPalette } from "../../core/utils/is-valid-color-palette.js";
import { DEFAULT_SCALE_FALLBACK_COLOR } from "../constants.js";
function isColorToken(value) {
  return !isNil(value) && typeof value !== "string";
}
function isColorPalette(value) {
  return typeof value === "string";
}
function getColorRuleResult(hasColorRules, colorRules, seriesForColorRule, categoryData, dimensionKey) {
  if (!hasColorRules) {
    return null;
  }
  const seriesEntry = seriesForColorRule.find(
    (s) => dimensionKey ? s._originalData === categoryData && s._dimensionKey === dimensionKey : s._originalData === categoryData && !s._dimensionKey
  );
  return seriesEntry ? colorRules.for(seriesEntry) : null;
}
function determineColorForDimension(colorRuleResult, originalPaletteColor, dimensionKey, paletteColorPosition) {
  if (isColorToken(colorRuleResult)) {
    return {
      color: colorRuleResult,
      legendColor: originalPaletteColor
    };
  }
  if (isColorPalette(colorRuleResult)) {
    return {
      color: getColorPalette(
        dimensionKey,
        paletteColorPosition,
        colorRuleResult
      ),
      legendColor: originalPaletteColor
    };
  }
  return {
    color: originalPaletteColor,
    legendColor: void 0
  };
}
function getDimensionColor({
  colorByDimension,
  categoryColor,
  categoryName,
  paletteColorPosition,
  colorPalette,
  colorPaletteMode,
  colorRuleResult
}) {
  const cachedColor = colorByDimension.get(categoryName);
  if (cachedColor) {
    return cachedColor;
  }
  if (categoryColor) {
    return { Default: categoryColor };
  }
  const appliedColorRule = colorRuleResult;
  if (typeof appliedColorRule === "string") {
    return getColorPalette(
      categoryName,
      paletteColorPosition,
      appliedColorRule,
      colorPaletteMode
    );
  }
  if (!isNull(appliedColorRule) && typeof appliedColorRule === "object") {
    return appliedColorRule;
  }
  if (colorPalette) {
    return getColorPalette(
      categoryName,
      paletteColorPosition,
      colorPalette,
      colorPaletteMode
    );
  }
  return { Default: DEFAULT_SCALE_FALLBACK_COLOR };
}
function handleSingleDimension({
  categoryData,
  rest,
  categoryName,
  categoryColor,
  colorByDimension,
  colorRules,
  colorPalette,
  colorPaletteMode,
  paletteColorPosition,
  uniqueDimensions,
  dimensionMap,
  newCategory,
  seriesForColorRule,
  hasColorRules
}) {
  const value = categoryData.value;
  const dimensionName = categoryName;
  const id = crypto.randomUUID();
  const colorRuleResult = getColorRuleResult(
    hasColorRules,
    colorRules,
    seriesForColorRule,
    categoryData
  );
  const color = getDimensionColor({
    colorByDimension,
    categoryColor,
    categoryName,
    paletteColorPosition,
    colorPalette,
    colorPaletteMode,
    colorRuleResult
  });
  colorByDimension.set(dimensionName, color);
  const newDimension = {
    ...rest,
    id,
    name: dimensionName,
    value
  };
  const dimensionMetadata = {
    id,
    name: dimensionName,
    value,
    color
  };
  uniqueDimensions.add(dimensionName);
  dimensionMap.set(newDimension, dimensionMetadata);
  newCategory.dimensions.push(newDimension);
}
function handleMultipleDimensions({
  categoryData,
  rest,
  categoryName,
  colorByDimension,
  colorByCategory,
  colorRules,
  colorPalette,
  categoricalColorPalette,
  uniqueDimensions,
  dimensionMap,
  newCategory,
  seriesForColorRule,
  hasColorRules
}) {
  Object.entries(categoryData.value).forEach(([dimensionKey, value], index) => {
    const colorRuleCategoryData = getColorRuleResult(
      hasColorRules,
      colorRules,
      seriesForColorRule,
      categoryData,
      dimensionKey
    );
    const paletteColorPosition = categoricalColorPalette ? index : 0;
    const colorByCategoryName = colorByCategory.get(categoryName);
    const dimensionColor = colorByCategoryName?.[dimensionKey];
    let originalPaletteColor;
    if (dimensionColor) {
      originalPaletteColor = dimensionColor;
    } else if (colorPalette && (isValidColorPalette(colorPalette) || isCustomColorPalette(colorPalette))) {
      originalPaletteColor = getColorPalette(
        dimensionKey,
        paletteColorPosition,
        colorPalette
      );
    } else {
      originalPaletteColor = { Default: DEFAULT_SCALE_FALLBACK_COLOR };
    }
    const { color, legendColor } = determineColorForDimension(
      colorRuleCategoryData,
      originalPaletteColor,
      dimensionKey,
      paletteColorPosition
    );
    if (!colorByDimension.has(dimensionKey)) {
      colorByDimension.set(dimensionKey, legendColor ?? color);
    }
    const categoryColorByCategory = colorByCategory.get(categoryName) ?? {};
    categoryColorByCategory[dimensionKey] = legendColor ?? color;
    colorByCategory.set(categoryName, categoryColorByCategory);
    const id = crypto.randomUUID();
    const name = dimensionKey;
    const newDimension = {
      ...rest,
      id,
      name,
      value,
      category: categoryName
    };
    const dimensionMetadata = {
      id,
      name,
      value,
      color,
      legendColor
    };
    uniqueDimensions.add(dimensionKey);
    dimensionMap.set(newDimension, dimensionMetadata);
    newCategory.dimensions.push(newDimension);
  });
}
export {
  getDimensionColor,
  handleMultipleDimensions,
  handleSingleDimension
};
