// @dynatrace/strato-components/esm/editors/base-code-editor/auto-completion/autocompletion.extension.js
import {
  startCompletion
} from "@codemirror/autocomplete";
import { render } from "../utils/render-node.js";
async function applyAndReopen(view, from, to, completion, apply, reopen) {
  if (apply === void 0 || typeof apply === "string") {
    const value = apply || completion.label;
    view.dispatch({
      changes: { from, to, insert: value },
      selection: { anchor: from + value.length },
      userEvent: "input.complete"
    });
  } else {
    apply(view, completion, from, to);
  }
  if (typeof reopen === "function" ? await reopen(completion) : reopen) {
    startCompletion(view);
  }
}
const autocompletionExtension = (autocompleteProvider, intlProps) => {
  return async (ctx) => {
    const result = await autocompleteProvider(ctx);
    if (!result) {
      return null;
    }
    return {
      ...result,
      options: result.suggestions.map((s, i) => ({
        ...s,
        info: s.popup ? (completion) => render(completion, s.popup, intlProps) : void 0,
        apply: (view, completion, from, to) => applyAndReopen(
          view,
          from,
          to,
          completion,
          s.apply,
          result.reopen ?? false
        ),
        ...result.keepOrder && { boost: 99 - i }
      }))
    };
  };
};
export {
  autocompletionExtension
};
