// @dynatrace/strato-components/charts/xy-chart/builders/XYChartImperativeHandlerBuilder.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var XYChartImperativeHandlerBuilder_exports = {};
__export(XYChartImperativeHandlerBuilder_exports, {
  XYChartImperativeHandlerBuilder: () => XYChartImperativeHandlerBuilder
});
module.exports = __toCommonJS(XYChartImperativeHandlerBuilder_exports);
var import_react = require("react");
var import_useGridPlotSize = require("../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_types = require("../../core/components/interactions-overlay/types.js");
var import_useTriggerZoom = require("../../core/components/zoom/hooks/useTriggerZoom.js");
var import_chart_types = require("../../core/types/chart-types.js");
var import_zoom = require("../../core/types/zoom.js");
var import_useApplyZoomDelta = require("../components/overlay/hooks/useApplyZoomDelta.js");
var import_useResetInteractions = require("../components/overlay/hooks/useResetInteractions.js");
var import_useDownload = require("../hooks/useDownload.js");
var import_useXYChartData = require("../hooks/useXYChartData.js");
var import_reduce_interaction_mode = require("../store/reduce-interaction-mode.js");
var import_selectors = require("../store/selectors.js");
var import_build_and_download_csv = require("../utils/build-and-download-csv.js");
const XYChartImperativeHandlerBuilder = ({
  chartRef,
  containerRef,
  config,
  children
}) => {
  const { seriesWithDatapoints, datapointMetadataMap, seriesToMetadata } = (0, import_useXYChartData.useXYChartData)();
  const { downloadDataCallback } = (0, import_useDownload.useDownload)();
  const setState = (0, import_selectors.useSetState)();
  const { zoomInEnabled, zoomOutEnabled } = (0, import_selectors.useZoomState)();
  const { axesMetadata } = (0, import_useXYChartData.useXYChartData)();
  const { width } = (0, import_useGridPlotSize.useGridPlotSize)();
  const applyZoom = (0, import_useApplyZoomDelta.useApplyZoomDelta)();
  const triggerZoom = (0, import_useTriggerZoom.useTriggerZoom)(width, applyZoom);
  const resetInteractions = (0, import_useResetInteractions.useResetInteractions)();
  (0, import_react.useImperativeHandle)(chartRef, () => ({
    element: containerRef.current,
    getConfig: () => {
      const { tooltip, ...exportableConfig } = config;
      return JSON.stringify({
        type: import_chart_types.ChartType.XYChart,
        ...exportableConfig,
        tooltip: { hidden: !!tooltip?.hidden }
      });
    },
    downloadData: () => {
      (0, import_build_and_download_csv.buildAndDownloadCsv)(
        seriesWithDatapoints,
        seriesToMetadata,
        datapointMetadataMap,
        downloadDataCallback
      );
    },
    exploreMode: () => {
      setState((0, import_reduce_interaction_mode.reduceInteractionMode)(import_types.INTERACTION_MODE.Explore));
    },
    zoomMode: () => {
      if (axesMetadata.isZoomEnabled) {
        setState((0, import_reduce_interaction_mode.reduceInteractionMode)(import_types.INTERACTION_MODE.Zoom));
      }
    },
    panMode: () => {
      if (zoomOutEnabled && axesMetadata.isPanEnabled) {
        setState((0, import_reduce_interaction_mode.reduceInteractionMode)(import_types.INTERACTION_MODE.Pan));
      }
    },
    zoomIn: () => {
      if (zoomInEnabled) {
        triggerZoom(import_zoom.ZOOM_CHANGE_TYPE.ZoomIn);
      }
    },
    zoomOut: () => {
      if (zoomOutEnabled) {
        triggerZoom(import_zoom.ZOOM_CHANGE_TYPE.ZoomOut);
      }
    },
    reset: () => {
      resetInteractions();
    }
  }));
  return children;
};
