// @dynatrace/strato-components/esm/charts/gauge/components/GaugeLabels.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { isNull, isUndefined } from "lodash-es";
import Colors from "@dynatrace/strato-design-tokens/colors";
import { replaceSlashedZero } from "../../core/typographyStyles.sty.js";
import { useDefaultLabels } from "../hooks/useDefaultLabels.js";
import { useGaugeChartLabelsContext } from "../hooks/useGaugeChartLabelsContext.js";
import { useGaugeChartMinMaxLabelsPosition } from "../hooks/useGaugeChartMinMaxLabelsPosition.js";
import { GaugeLabel } from "./GaugeLabel/GaugeLabel.js";
import { GaugeLabelsWrapper } from "./GaugeLabelsWrapper/GaugeLabelsWrapper.js";
const GaugeLabels = ({ value, prefixIcon }) => {
  const { label } = useGaugeChartLabelsContext();
  const { defaultLabel, defaultMaxLabel, defaultMinLabel, defaultValueLabel } = useDefaultLabels(value);
  const { maxLabelPosition, minLabelPosition, minMaxPadding } = useGaugeChartMinMaxLabelsPosition();
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(
      "g",
      {
        fill: Colors.Text.Neutral.Default,
        id: "internal-labels",
        className: replaceSlashedZero,
        children: /* @__PURE__ */ jsxs(GaugeLabelsWrapper, { children: [
          /* @__PURE__ */ jsx(
            GaugeLabel,
            {
              id: "value-label",
              value: defaultValueLabel,
              prefixIcon,
              responsive: true
            }
          ),
          !isUndefined(label) && /* @__PURE__ */ jsx(
            GaugeLabel,
            {
              id: "label",
              value: isNull(label) ? defaultLabel : label,
              responsive: true
            }
          )
        ] })
      }
    ),
    /* @__PURE__ */ jsxs("g", { fill: Colors.Text.Neutral.Default, id: "min-max-labels", children: [
      /* @__PURE__ */ jsx(
        GaugeLabel,
        {
          value: defaultMinLabel,
          x: minLabelPosition.x,
          y: minLabelPosition.y + minMaxPadding,
          autoHide: true
        }
      ),
      /* @__PURE__ */ jsx(
        GaugeLabel,
        {
          value: defaultMaxLabel,
          x: maxLabelPosition.x,
          y: maxLabelPosition.y + minMaxPadding,
          autoHide: true
        }
      )
    ] })
  ] });
};
export {
  GaugeLabels
};
