// @dynatrace/strato-components/charts/xy-chart/XYChartRoot.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var XYChartRoot_exports = {};
__export(XYChartRoot_exports, {
  XYChart: () => XYChart,
  XYChartAnnotations: () => XYChartAnnotations,
  XYChartRoot: () => XYChartRoot
});
module.exports = __toCommonJS(XYChartRoot_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_XYChartLegend = require("./components/legend/XYChartLegend.js");
var import_defaults = require("./constants/defaults.js");
var import_XYChartAxesConfig_context = require("./context/axes/XYChartAxesConfig.context.js");
var import_XYChartAnnotationsConfig_context = require("./context/XYChartAnnotationsConfig.context.js");
var import_XYChartColorPalette_context = require("./context/XYChartColorPalette.context.js");
var import_XYChartLegendHidden_context = require("./context/XYChartLegendHidden.context.js");
var import_XYChartThresholds_context = require("./context/XYChartThresholds.context.js");
var import_XYChartToolbar_context = require("./context/XYChartToolbar.context.js");
var import_XYChartTruncationMode_context = require("./context/XYChartTruncationMode.context.js");
var import_useXYChartConfig = require("./hooks/useXYChartConfig.js");
var import_XYChartAnnotationsDataProvider = require("./providers/XYChartAnnotationsDataProvider.js");
var import_XYChartAnnotationsShownIndicatorsProvider = require("./providers/XYChartAnnotationsShownIndicatorsProvider.js");
var import_XYChartColorScalesProvider = require("./providers/XYChartColorScalesProvider.js");
var import_XYChartDataProvider = require("./providers/XYChartDataProvider.js");
var import_XYChartDownloadProvider = require("./providers/XYChartDownloadProvider.js");
var import_XYChartIdProvider = require("./providers/XYChartIdProvider.js");
var import_XYChartStoreDomainSync = require("./providers/XYChartStoreDomainSync.js");
var import_XYChartStoreProvider = require("./providers/XYChartStoreProvider.js");
var import_XYChartVisibleAxesProvider = require("./providers/XYChartVisibleAxesProvider.js");
var import_Annotations = require("./slots/Annotations/Annotations.js");
var import_Marker = require("./slots/Annotations/Marker.js");
var import_Tooltip = require("./slots/Annotations/Tooltip.js");
var import_Track = require("./slots/Annotations/Track.js");
var import_AreaSeries = require("./slots/AreaSeries.js");
var import_BarSeries = require("./slots/BarSeries.js");
var import_CustomTooltip = require("./slots/CustomTooltip.js");
var import_DotSeries = require("./slots/DotSeries.js");
var import_Legend = require("./slots/Legend.js");
var import_LineSeries = require("./slots/LineSeries.js");
var import_RectSeries = require("./slots/RectSeries.js");
var import_Select = require("./slots/Select.js");
var import_ThresholdIndicator = require("./slots/ThresholdIndicator.js");
var import_Toolbar = require("./slots/Toolbar.js");
var import_Tooltip2 = require("./slots/Tooltip.js");
var import_XAxis = require("./slots/XAxis.js");
var import_YAxis = require("./slots/YAxis.js");
var import_is_empty_data = require("./utils/is-empty-data.js");
var import_iterate_slots = require("./utils/iterate-slots.js");
var import_xy_chart_configuration = require("./utils/xy-chart-configuration.js");
var import_XYChart = require("./XYChart.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_Intent = require("../../core/slots/Intent/Intent.js");
var import_ChartLayout = require("../core/components/chart-layout/ChartLayout.js");
var import_useAutoLegendRefresh = require("../core/components/chart-layout/hooks/useAutoLegendRefresh.js");
var import_get_chart_states_template = require("../core/components/chart-states/utils/get-chart-states-template.js");
var import_OverlayContext = require("../core/components/overlay-tooltip/contexts/OverlayContext.js");
var import_TooltipContext = require("../core/components/overlay-tooltip/contexts/TooltipContext.js");
var import_ColorRuleContext = require("../core/context/ColorRuleContext.js");
var import_EmptyStateSlot = require("../core/slots/chart-states/EmptyStateSlot.js");
var import_ErrorStateSlot = require("../core/slots/chart-states/ErrorStateSlot.js");
var import_ColorRule = require("../core/slots/ColorRule.js");
var import_coerceSizeValue = require("../core/utils/coerceSizeValue.js");
const COMPONENT_NAME_BEHAVIOURAL_TRACKING = "XYChart";
const XYChartRoot = /* @__PURE__ */ (0, import_react.forwardRef)((props, ref) => {
  const {
    data,
    children,
    height,
    width,
    loading = import_defaults.DEFAULT_LOADING_STATE,
    style,
    className,
    __downloadDataCallback: downloadDataCallback,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow
  } = props;
  const [behaviorTrackingProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(props);
  const slots = (0, import_iterate_slots.iterateSlots)(props.children);
  const config = (0, import_useXYChartConfig.useXYChartConfig)(props, slots);
  const { emptyState, errorState } = (0, import_get_chart_states_template.getChartStatesTemplates)(props.children);
  const containerRef = (0, import_react.useRef)(null);
  const rootRef = (0, import_useMergeRefs.useMergeRefs)([
    ref,
    containerRef
  ]);
  const {
    hidden: legendHidden,
    // eslint-disable-next-line @typescript-eslint/no-deprecated
    ratio,
    initialRatio,
    onRatioChange,
    position: legendPosition
  } = config.legend;
  const legendRatio = initialRatio ?? ratio;
  const chartHeight = (0, import_coerceSizeValue.coerceSizeValue)(height ?? import_defaults.DEFAULT_CHART_HEIGHT);
  const internalConfig = (0, import_xy_chart_configuration.buildInternalConfig)(config, slots);
  const chartLayoutRef = (0, import_useAutoLegendRefresh.useAutoLegendRefresh)(data);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartIdProvider.XYChartIdProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartThresholds_context.XYChartThresholdsContext.Provider, { value: internalConfig.thresholds, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartAxesConfig_context.XYChartAxesConfigContext.Provider, { value: internalConfig.axes, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartColorPalette_context.XYChartColorPaletteContext.Provider, { value: config.colorPalette, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_XYChartDownloadProvider.XYChartDownloadProvider,
    {
      downloadDataCallback,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartToolbar_context.XYChartToolbarContext.Provider, { value: config.toolbar, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_XYChartTruncationMode_context.XYChartTruncationModeContext.Provider,
        {
          value: { truncationMode: config.truncationMode },
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_XYChartLegendHidden_context.XYChartLegendHiddenContext.Provider,
            {
              value: Boolean(legendHidden),
              children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_XYChartAnnotationsConfig_context.XYChartAnnotationsConfig.Provider,
                {
                  value: config.annotations,
                  children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ColorRuleContext.ColorRuleContext.Provider, { value: config.colorRules, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartStoreProvider.XYChartStoreProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                    import_XYChartDataProvider.XYChartChartDataProvider,
                    {
                      data,
                      seriesVariantsProps: internalConfig.series,
                      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartStoreDomainSync.XYChartStoreDomainSync, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartAnnotationsDataProvider.XYChartAnnotationsDataProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartVisibleAxesProvider.XYChartVisibleAxesProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartColorScalesProvider.XYChartColorScalesProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartAnnotationsShownIndicatorsProvider.XYChartAnnotationsShownIndicatorsProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayContext.OverlayProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TooltipContext.TooltipProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
                        import_ChartLayout.ChartLayout,
                        {
                          ref: chartLayoutRef,
                          chartHeight,
                          chartWidth: width,
                          className,
                          emptyState,
                          errorState,
                          isEmpty: (0, import_is_empty_data.isEmptyData)(
                            children,
                            data
                          ),
                          showLoader: loading,
                          style,
                          "data-dtrum-mask": dataDtrumMask,
                          "data-dtrum-allow": dataDtrumAllow,
                          ...behaviorTrackingProps,
                          "data-dt-component": COMPONENT_NAME_BEHAVIOURAL_TRACKING,
                          children: [
                            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ChartLayout.ChartLayout.Graph, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                              import_XYChart.XYChart,
                              {
                                config,
                                ref: rootRef
                              }
                            ) }),
                            !legendHidden && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                              import_ChartLayout.ChartLayout.Legend,
                              {
                                position: legendPosition,
                                initialRatio: legendRatio,
                                onResize: onRatioChange,
                                children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartLegend.XYChartLegend, {})
                              }
                            )
                          ]
                        }
                      ) }) }) }) }) }) }) })
                    }
                  ) }) })
                }
              )
            }
          )
        }
      ) })
    }
  ) }) }) }) });
});
const XYChart = Object.assign(XYChartRoot, {
  Annotations: import_Annotations.Annotations,
  RectSeries: import_RectSeries.RectSeries,
  DotSeries: import_DotSeries.DotSeries,
  LineSeries: import_LineSeries.LineSeries,
  BarSeries: import_BarSeries.BarSeries,
  AreaSeries: import_AreaSeries.AreaSeries,
  XAxis: import_XAxis.XAxis,
  YAxis: import_YAxis.YAxis,
  Legend: import_Legend.Legend,
  Toolbar: import_Toolbar.Toolbar,
  /** @deprecated Use `XYChart.Tooltip` instead. */
  // eslint-disable-next-line @typescript-eslint/no-deprecated
  CustomTooltip: import_CustomTooltip.CustomTooltip,
  Tooltip: import_Tooltip2.XYChartTooltipSlot,
  Intent: import_Intent.Intent,
  Select: import_Select.Select,
  ThresholdIndicator: import_ThresholdIndicator.ThresholdIndicator,
  EmptyState: import_EmptyStateSlot.EmptyStateSlot,
  ErrorState: import_ErrorStateSlot.ErrorStateSlot,
  ColorRule: import_ColorRule.ColorRule
});
XYChart.displayName = "XYChart";
const XYChartAnnotations = {
  Marker: import_Marker.Marker,
  Track: import_Track.Track,
  Tooltip: import_Tooltip.AnnotationsTooltip
};
