// @dynatrace/strato-components/charts/xy-chart/XYChart.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var XYChart_exports = {};
__export(XYChart_exports, {
  XYChart: () => XYChart
});
module.exports = __toCommonJS(XYChart_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_XYChartImperativeHandlerBuilder = require("./builders/XYChartImperativeHandlerBuilder.js");
var import_XYChartAnnotationsArea = require("./components/annotations/XYChartAnnotationsArea.js");
var import_XYChartAnnotationsIndicators = require("./components/annotations/XYChartAnnotationsIndicators.js");
var import_XYChartBottomAxisArea = require("./components/axes/XYChartBottomAxisArea.js");
var import_XYChartLeftAxisArea = require("./components/axes/XYChartLeftAxisArea.js");
var import_XYChartRightAxisArea = require("./components/axes/XYChartRightAxisArea.js");
var import_XYChartTopAxisArea = require("./components/axes/XYChartTopAxisArea.js");
var import_BarLabelLayout = require("./components/displays/BarDisplay/BarLabelLayout.js");
var import_XYChartGridLayout = require("./components/layout/XYChartGridLayout.js");
var import_XYChartOverlay = require("./components/overlay/XYChartOverlay.js");
var import_XYChartThresholdArea = require("./components/thresholds/XYChartThresholdArea.js");
var import_defaults = require("./constants/defaults.js");
var import_XYChartCustomTooltip_context = require("./context/XYChartCustomTooltip.context.js");
var import_XYChartSelectConfig_context = require("./context/XYChartSelectConfig.context.js");
var import_useVisibleAxesConfig = require("./hooks/useVisibleAxesConfig.js");
var import_useXYChartAnnotationsConfig = require("./hooks/useXYChartAnnotationsConfig.js");
var import_useXYChartAnnotationsData = require("./hooks/useXYChartAnnotationsData.js");
var import_useXYChartAnnotationsHiddenTracks = require("./hooks/useXYChartAnnotationsHiddenTracks.js");
var import_useXYChartData = require("./hooks/useXYChartData.js");
var import_XYChartAxisBorder = require("./layout/XYChartAxisBorder.js");
var import_XYChartLayout = require("./layout/XYChartLayout.js");
var import_XYChartProviders = require("./providers/XYChartProviders.js");
var import_get_padding_by_axes = require("./utils/get-padding-by-axes.js");
var import_grid_templates = require("./utils/grid-templates.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_defaultChartValues_css = require("../../core/styles/defaultChartValues.sty.js");
var import_get_annotations_visible_area = require("../core/components/annotations/utils/get-annotations-visible-area.js");
var import_ChartGridRoot = require("../core/components/chart-grid/ChartGridRoot.js");
var import_useChartLayoutLoading = require("../core/components/chart-layout/hooks/useChartLayoutLoading.js");
var import_useGraphSize = require("../core/components/chart-layout/hooks/useGraphSize.js");
var import_ChartLoaderOverlay = require("../core/components/chart-loader-overlay/ChartLoaderOverlay.js");
const XYChart = (0, import_react.forwardRef)(
  ({ config }, forwardedRef) => {
    const loading = (0, import_useChartLayoutLoading.useChartLayoutLoading)();
    const { width, height } = (0, import_useGraphSize.useGraphSize)();
    const chartRef = (0, import_react.useRef)(null);
    const chartAndForwardedRef = (0, import_useMergeRefs.useMergeRefs)([
      forwardedRef,
      chartRef
    ]);
    const { top, bottom, left, right } = (0, import_useVisibleAxesConfig.useVisibleAxesConfig)();
    const { axesMetadata } = (0, import_useXYChartData.useXYChartData)();
    const topAxisMaxSize = top?.tickLabelMaxSize;
    const bottomAxisMaxSize = bottom?.tickLabelMaxSize;
    const leftAxisMaxSize = left?.tickLabelMaxSize;
    const rightAxisMaxSize = right?.tickLabelMaxSize;
    const paddings = (0, import_get_padding_by_axes.getPaddingByAxes)({ top, bottom, left, right });
    const chartGridHeight = height - paddings.vertical();
    const chartGridWidth = width - paddings.horizontal();
    const annotationsData = (0, import_useXYChartAnnotationsData.useXYChartAnnotationsData)();
    const annotationsConfig = (0, import_useXYChartAnnotationsConfig.useXYChartAnnotationsConfig)();
    const tracksQuantity = annotationsData?.size ?? 0;
    const hiddenTracks = (0, import_useXYChartAnnotationsHiddenTracks.useXYChartAnnotationsHiddenTracks)();
    const annotationsVisibleArea = (0, import_get_annotations_visible_area.getAnnotationsVisibleArea)(
      tracksQuantity,
      annotationsConfig?.fixedTracks,
      hiddenTracks.length
    );
    const annotationsAreaHeight = annotationsVisibleArea + (tracksQuantity - hiddenTracks.length > 0 ? import_defaults.ANNOTATIONS_PADDING_BOTTOM : 0);
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "div",
      {
        className: import_defaultChartValues_css.defaultChartSize,
        style: {
          paddingTop: paddings.top,
          paddingLeft: paddings.left
        },
        "aria-label": import_defaults.XY_CHART_ARIA_LABEL,
        role: "img",
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_ChartGridRoot.ChartGrid,
          {
            width: chartGridWidth,
            height: chartGridHeight,
            gridTemplateAreas: import_grid_templates.gridTemplateAreas,
            gridTemplateColumns: (0, import_grid_templates.gridTemplateColumns)(
              leftAxisMaxSize,
              rightAxisMaxSize
            ),
            gridTemplateRows: (0, import_grid_templates.gridTemplateRows)(
              bottomAxisMaxSize,
              topAxisMaxSize,
              annotationsAreaHeight
            ),
            ref: chartAndForwardedRef,
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartCustomTooltip_context.XYChartCustomTooltipContext.Provider, { value: config.tooltip, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartSelectConfig_context.XYChartSelectConfigContext.Provider, { value: config.select, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartProviders.XYChartProviders, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
              import_XYChartImperativeHandlerBuilder.XYChartImperativeHandlerBuilder,
              {
                chartRef: forwardedRef,
                containerRef: chartRef,
                config,
                children: [
                  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartTopAxisArea.XYChartTopAxisArea, {}),
                  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartBottomAxisArea.XYChartBottomAxisArea, {}),
                  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartLeftAxisArea.XYChartLeftAxisArea, {}),
                  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartRightAxisArea.XYChartRightAxisArea, {}),
                  !!tracksQuantity && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartAnnotationsArea.XYChartAnnotationsArea, {}),
                  /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_ChartGridRoot.ChartGrid.PlotArea, { children: [
                    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartGridLayout.XYChartGridLayout, {}),
                    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartLayout.XYChartLayout, {}),
                    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_BarLabelLayout.BarLabelLayout, {}),
                    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartAxisBorder.XYChartAxisBorder, {}),
                    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartAnnotationsIndicators.XYChartAnnotationsIndicators, {}),
                    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                      import_ChartLoaderOverlay.ChartLoaderOverlay,
                      {
                        loading,
                        zoomEnabled: axesMetadata.isZoomEnabled,
                        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartOverlay.XYChartOverlay, {})
                      }
                    )
                  ] }),
                  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartThresholdArea.XYChartThresholdArea, {})
                ]
              }
            ) }) }) })
          }
        )
      }
    );
  }
);
XYChart.displayName = "XYChart";
