// @dynatrace/strato-components/charts/xy-chart/components/thresholds/builder.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var builder_exports = {};
__export(builder_exports, {
  buildThresholds: () => buildThresholds
});
module.exports = __toCommonJS(builder_exports);
var import_get_y_value = require("./get-y-value.js");
var import_type_guards = require("./type-guards.js");
var import_utils = require("./utils.js");
var import_constants = require("../../../core/components/threshold-bar/types/constants.js");
var import_type_guards2 = require("../../../core/types/type-guards.js");
var import_getValidColor = require("../../../core/utils/getValidColor.js");
var import_XYChartThresholds_context = require("../../context/XYChartThresholds.context.js");
var import_cast_data_types = require("../../processing-strategies/utils/cast-data/cast-data-types.js");
var import_guards = require("../../types/guards.js");
function getStrokeOnly(threshold) {
  const firstData = (0, import_type_guards.isThresholdDynamic)(threshold) ? threshold.data[0] : threshold.data;
  const firstMarkerDomain = {
    y0: (0, import_get_y_value.getYValue)(firstData, "min"),
    y1: (0, import_get_y_value.getYValue)(firstData, "max")
  };
  return ((0, import_type_guards2.isPointThreshold)(firstMarkerDomain.y0, firstMarkerDomain.y1) || threshold.strokeOnly) ?? false;
}
function hasValidAxes(threshold, yAxisMetadata, xAxisMetadata = void 0) {
  let xAxisValid = true;
  if ((0, import_type_guards.isThresholdDynamic)(threshold) && xAxisMetadata) {
    xAxisValid = threshold.xAxisId !== void 0 && threshold.xAxisId in xAxisMetadata;
  }
  return xAxisValid && threshold.yAxisId in yAxisMetadata;
}
const buildThresholds = (yAxes, xAxes, thresholdSlots) => {
  const xAxisMetadata = xAxes.filter(import_guards.isAnyNumericalAxis).reduce((acc, axis) => {
    acc[axis.id] = {
      position: axis.position,
      type: axis.type,
      reversed: axis.reversed
    };
    return acc;
  }, {});
  const yAxisMetadata = yAxes.filter(import_guards.isAnyNumericalAxis).reduce((acc, axis) => {
    acc[axis.id] = {
      position: axis.position,
      type: axis.type
    };
    return acc;
  }, {});
  const leftAxisWithThresholds = /* @__PURE__ */ new Set();
  const rightAxisWithThresholds = /* @__PURE__ */ new Set();
  const thresholdsData = thresholdSlots.reduce(
    (acc, threshold) => {
      if (hasValidAxes(threshold, yAxisMetadata, xAxisMetadata)) {
        const yAxis = yAxisMetadata[threshold.yAxisId];
        const baseThresholdData = {
          id: crypto.randomUUID(),
          yAxisId: threshold.yAxisId,
          color: (0, import_getValidColor.getValidColor)(threshold.color),
          position: yAxis.position,
          label: threshold.label,
          strokeOnly: getStrokeOnly(threshold)
        };
        if ((0, import_utils.axisPositionIncludes)("left")(yAxis.position)) {
          leftAxisWithThresholds.add(threshold.yAxisId);
        }
        if ((0, import_utils.axisPositionIncludes)("right")(yAxis.position)) {
          rightAxisWithThresholds.add(threshold.yAxisId);
        }
        if ((0, import_type_guards.isThresholdDynamic)(threshold)) {
          const xAxis = xAxisMetadata[threshold.xAxisId];
          const castedData = (0, import_cast_data_types.castDynamicThresholdsDataType)(
            threshold.data,
            xAxis.type,
            yAxis.type
          );
          if (castedData.length) {
            const dynamicData = xAxis.reversed ? castedData.toReversed() : castedData;
            acc.push({
              ...baseThresholdData,
              xAxisId: threshold.xAxisId,
              data: dynamicData
            });
          }
        } else {
          const castedData = (0, import_cast_data_types.castStaticThresholdsDataType)(
            threshold.data,
            yAxis.type
          );
          if (castedData) {
            acc.push({
              ...baseThresholdData,
              data: castedData
            });
          }
        }
      }
      return acc;
    },
    []
  );
  if (!thresholdsData.length) {
    return import_XYChartThresholds_context.NoThresholdData;
  }
  const thresholds = {
    thresholds: thresholdsData,
    none: false,
    offset: {
      left: (0, import_utils.axesPositionsIncludes)("left", thresholdsData) ? import_constants.THRESHOLD_BAR_THICKNESS : 0,
      right: (0, import_utils.axesPositionsIncludes)("right", thresholdsData) ? import_constants.THRESHOLD_BAR_THICKNESS : 0
    },
    yAxes: {
      left: leftAxisWithThresholds,
      right: rightAxisWithThresholds
    }
  };
  return thresholds;
};
