// @dynatrace/strato-components/charts/xy-chart/components/thresholds/utils.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var utils_exports = {};
__export(utils_exports, {
  axesPositionsIncludes: () => axesPositionsIncludes,
  axisPositionIncludes: () => axisPositionIncludes,
  getDynamicThresholdDomain: () => getDynamicThresholdDomain,
  mapIndicatorDataFrom: () => mapIndicatorDataFrom,
  markersReducer: () => markersReducer
});
module.exports = __toCommonJS(utils_exports);
var import_d3_scale = require("d3-scale");
var import_lodash_es = require("lodash");
var import_get_y_value = require("./get-y-value.js");
var import_type_guards = require("./type-guards.js");
var import_get_marker_value = require("../../../core/components/threshold-bar/utils/get-marker-value.js");
var import_get_threshold_pixel_ts = require("../../../core/components/threshold-bar/utils/get-threshold-pixel.ts.js");
var import_messages = require("../../../core/messages.js");
var import_get_scale = require("../../utils/get-scale.js");
var import_get_axis_formatter_with_time_fallback = require("../overlay/utils/get-axis-formatter-with-time-fallback.js");
function getDynamicThresholdDomain(data, side, visibleDomain) {
  const minXPoint = findPointWithMinX(data);
  const maxXPoint = findPointWithMaxX(data);
  const targetXDomain = side === "left" ? visibleDomain.min : visibleDomain.max;
  if (targetXDomain <= minXPoint.x.valueOf()) {
    return getYRangeFromPoint(minXPoint);
  }
  if (targetXDomain >= maxXPoint.x.valueOf()) {
    return getYRangeFromPoint(maxXPoint);
  }
  return interpolateYRange(data, targetXDomain);
}
function findPointWithMinX(data) {
  return (0, import_lodash_es.minBy)(data, (point) => point.x.valueOf()) ?? data[0];
}
function findPointWithMaxX(data) {
  return (0, import_lodash_es.maxBy)(data, (point) => point.x.valueOf()) ?? data[data.length - 1];
}
function getYRangeFromPoint(point) {
  return {
    y0: (0, import_get_y_value.getYValue)(point, "min"),
    y1: (0, import_get_y_value.getYValue)(point, "max")
  };
}
function interpolateYRange(data, targetX) {
  const xValues = data.map((point) => Number(point.x));
  const minYValues = data.map((point) => (0, import_get_y_value.getYValue)(point, "min"));
  const maxYValues = data.map((point) => (0, import_get_y_value.getYValue)(point, "max"));
  const minYScale = (0, import_d3_scale.scaleLinear)(xValues, minYValues).clamp(true);
  const maxYScale = (0, import_d3_scale.scaleLinear)(xValues, maxYValues).clamp(true);
  return {
    y0: minYScale(targetX),
    y1: maxYScale(targetX)
  };
}
function markersReducer({
  axis,
  axisConfig,
  axesIdToScales,
  axesIdToFormatter,
  width,
  axes
}) {
  return (acc, threshold) => {
    if (!axes.find((axis2) => axis2.id === threshold.yAxisId)) {
      return acc;
    }
    if (axisPositionIncludes(axis)(threshold.position)) {
      const yScale = (0, import_get_scale.getCurrentScale)(
        axesIdToScales,
        threshold.yAxisId
      );
      if ((0, import_type_guards.isThresholdDynamic)(threshold)) {
        const xScale = (0, import_get_scale.getCurrentScale)(
          axesIdToScales,
          threshold.xAxisId
        );
        const data = threshold.data;
        const minElement = (0, import_lodash_es.minBy)(data, (e) => (0, import_get_y_value.getYValue)(e, "min"));
        const maxElement = (0, import_lodash_es.maxBy)(data, (e) => (0, import_get_y_value.getYValue)(e, "max"));
        if (!minElement || !maxElement) {
          return acc;
        }
        acc.push({
          ...threshold,
          y0: (0, import_get_threshold_pixel_ts.getThresholdPixel)((0, import_get_y_value.getYValue)(minElement, "min"), yScale),
          y1: (0, import_get_threshold_pixel_ts.getThresholdPixel)((0, import_get_y_value.getYValue)(maxElement, "max"), yScale),
          value: "dynamic",
          isShadow: true
        });
        const domain = xScale.domain();
        const xMinMax = {
          min: domain[0].valueOf(),
          max: domain[1].valueOf()
        };
        const { y0, y1 } = getDynamicThresholdDomain(data, axis, xMinMax);
        acc.push({
          ...threshold,
          y0: (0, import_get_threshold_pixel_ts.getThresholdPixel)(y0, yScale),
          y1: (0, import_get_threshold_pixel_ts.getThresholdPixel)(y1, yScale),
          value: import_messages.THRESHOLD.dynamic.defaultMessage
        });
      } else {
        const markerDomain = {
          y0: (0, import_get_y_value.getYValue)(threshold.data, "min"),
          y1: (0, import_get_y_value.getYValue)(threshold.data, "max")
        };
        acc.push({
          ...threshold,
          y0: (0, import_get_threshold_pixel_ts.getThresholdPixel)(markerDomain.y0, yScale),
          y1: (0, import_get_threshold_pixel_ts.getThresholdPixel)(markerDomain.y1, yScale),
          value: (0, import_get_marker_value.getMarkerValue)(
            markerDomain,
            (0, import_get_axis_formatter_with_time_fallback.getAxisFormatterWithTimeFallback)({
              axisConfig,
              axisId: threshold.yAxisId,
              scale: yScale,
              axesIdToFormatter,
              width
            })
          )
        });
      }
    }
    return acc;
  };
}
const axesPositionsIncludes = (axisPosition, axesPositions) => {
  const expectedPositionsIncludes = axisPositionIncludes(axisPosition);
  return !!axesPositions.find(
    ({ position }) => expectedPositionsIncludes(position)
  );
};
const axisPositionIncludes = (expectedPosition) => (actualPosition) => {
  return [expectedPosition, "both"].includes(actualPosition);
};
const mapIndicatorDataFrom = ({
  axisIdToScales,
  threshold,
  width
}) => {
  const yScale = (0, import_get_scale.getCurrentScale)(
    axisIdToScales,
    threshold.yAxisId
  );
  if ((0, import_type_guards.isThresholdDynamic)(threshold)) {
    const data = threshold.data;
    const xScale = (0, import_get_scale.getCurrentScale)(
      axisIdToScales,
      threshold.xAxisId
    );
    const initialSegment = {
      x0: 0,
      x1: 0,
      y0: (0, import_get_threshold_pixel_ts.getThresholdPixel)((0, import_get_y_value.getYValue)(data[0], "min"), yScale),
      y1: (0, import_get_threshold_pixel_ts.getThresholdPixel)((0, import_get_y_value.getYValue)(data[0], "max"), yScale)
    };
    const finalSegment = {
      x0: width,
      x1: width,
      y0: (0, import_get_threshold_pixel_ts.getThresholdPixel)((0, import_get_y_value.getYValue)(data[data.length - 1], "min"), yScale),
      y1: (0, import_get_threshold_pixel_ts.getThresholdPixel)((0, import_get_y_value.getYValue)(data[data.length - 1], "max"), yScale)
    };
    const segments = data.map((coordinate) => ({
      x0: (0, import_get_threshold_pixel_ts.getThresholdPixel)(coordinate.x, xScale),
      x1: (0, import_get_threshold_pixel_ts.getThresholdPixel)(coordinate.x, xScale),
      y0: (0, import_get_threshold_pixel_ts.getThresholdPixel)((0, import_get_y_value.getYValue)(coordinate, "min"), yScale),
      y1: (0, import_get_threshold_pixel_ts.getThresholdPixel)((0, import_get_y_value.getYValue)(coordinate, "max"), yScale)
    }));
    return [initialSegment, ...segments, finalSegment];
  } else {
    const data = threshold.data;
    const yMinScaleValue = (0, import_get_threshold_pixel_ts.getThresholdPixel)((0, import_get_y_value.getYValue)(data, "min"), yScale);
    const yMaxScaleValue = (0, import_get_threshold_pixel_ts.getThresholdPixel)((0, import_get_y_value.getYValue)(data, "max"), yScale);
    return [
      {
        x0: 0,
        x1: 0,
        y0: yMinScaleValue,
        y1: yMaxScaleValue
      },
      {
        x0: width,
        x1: width,
        y0: yMinScaleValue,
        y1: yMaxScaleValue
      }
    ];
  }
};
