// @dynatrace/strato-components/charts/xy-chart/components/thresholds/XYChartThresholdArea.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var XYChartThresholdArea_exports = {};
__export(XYChartThresholdArea_exports, {
  XYChartThresholdArea: () => XYChartThresholdArea
});
module.exports = __toCommonJS(XYChartThresholdArea_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_get_border_threshold = require("./get-border-threshold.js");
var import_utils = require("./utils.js");
var import_ChartGridRoot = require("../../../core/components/chart-grid/ChartGridRoot.js");
var import_useGridPlotSize = require("../../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_ThresholdBar = require("../../../core/components/threshold-bar/ThresholdBar.js");
var import_useAxesFormatter = require("../../hooks/useAxesFormatter.js");
var import_useGetThresholds = require("../../hooks/useGetThresholds.js");
var import_useVisibleAxesConfig = require("../../hooks/useVisibleAxesConfig.js");
var import_useXYChartData = require("../../hooks/useXYChartData.js");
var import_useXYChartScales = require("../../hooks/useXYChartScales.js");
var import_selectors = require("../../store/selectors.js");
var import_grid_templates = require("../../utils/grid-templates.js");
const XYChartThresholdArea = () => {
  const { thresholds, none } = (0, import_useGetThresholds.useGetThresholds)();
  const { left, right } = (0, import_useVisibleAxesConfig.useVisibleAxesConfig)();
  const axesIdToScales = (0, import_useXYChartScales.useXYChartScales)();
  const axesIdToFormatter = (0, import_useAxesFormatter.useAxesFormatter)();
  const { width } = (0, import_useGridPlotSize.useGridPlotSize)();
  const { left: leftConfig, right: rightConfig } = (0, import_useVisibleAxesConfig.useVisibleAxesConfig)();
  const setState = (0, import_selectors.useSetState)();
  const { axes } = (0, import_useXYChartData.useXYChartData)();
  if (none) {
    return null;
  }
  const leftMarkers = thresholds.reduce(
    (0, import_utils.markersReducer)({
      axis: "left",
      axisConfig: leftConfig,
      axesIdToScales,
      axesIdToFormatter,
      width,
      axes
    }),
    []
  );
  const rightMarkers = thresholds.reduce(
    (0, import_utils.markersReducer)({
      axis: "right",
      axisConfig: rightConfig,
      axesIdToScales,
      axesIdToFormatter,
      width,
      axes
    }),
    []
  );
  const handleHover = (id) => {
    setState((prev) => ({
      ...prev,
      thresholds: {
        ...prev.thresholds,
        hoveredThresholdId: id
      }
    }));
  };
  const { leftBorder, rightBorder } = (0, import_get_border_threshold.getBorderThreshold)(left, right);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    leftMarkers.length ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_ChartGridRoot.ChartGrid.GridArea,
      {
        gridArea: import_grid_templates.AREAS.leftThreshold,
        "data-testid": "left-threshold-area",
        children: (args) => {
          return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_ThresholdBar.ThresholdBar,
            {
              borderArea: leftBorder,
              markers: leftMarkers,
              length: args.availableHeight,
              layout: "vertical",
              onHover: handleHover
            }
          );
        }
      },
      "leftThreshold"
    ) : null,
    rightMarkers.length ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_ChartGridRoot.ChartGrid.GridArea,
      {
        gridArea: import_grid_templates.AREAS.rightThreshold,
        "data-testid": "right-threshold-area",
        children: (args) => {
          return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_ThresholdBar.ThresholdBar,
            {
              borderArea: rightBorder,
              markers: rightMarkers,
              length: args.availableHeight,
              layout: "vertical",
              onHover: handleHover
            }
          );
        }
      },
      "rightThreshold"
    ) : null
  ] });
};
XYChartThresholdArea["displayName"] = "XYChartThresholdArea";
