// @dynatrace/strato-components/esm/tables/DataTable/features/ColumnSizing/ColumnSizing.js
import { functionalUpdate } from "@tanstack/react-table";
import * as styles from "./ColumnSizing.sty.js";
import { useControllableState } from "../../../../core/hooks/useControllableState.js";
import { useDataTableConfigElement } from "../../../hooks/useDataTableConfig.js";
import {
  DATA_TABLE_DEFAULT_MIN_COLUMN_WIDTH,
  DATA_TABLE_RESIZING_LEFT_POSITION_CSS_VAR,
  DATA_TABLE_RESIZING_RIGHT_POSITION_CSS_VAR,
  DATA_TABLE_STRATO_RESIZING_SIZE_CSS_VAR
} from "../../constants.js";
import { isFractionWidth } from "../Layout/width-utils.js";
const COLUMN_RESIZE_KEYBOARD_OFFSET = 10;
const COLUMN_RESIZE_KEYBOARD_CTRL_MULTIPLIER = 10;
const getClientX = (event) => {
  if ("touches" in event && event.touches.length > 0) {
    return event.touches[0].clientX;
  }
  if ("changedTouches" in event && event.changedTouches.length > 0) {
    return event.changedTouches[0].clientX;
  }
  if ("clientX" in event) {
    return event.clientX;
  }
};
const getHeaderCellWidth = (resizeHandle) => resizeHandle.parentElement?.getBoundingClientRect().width;
const clearResizerFocus = () => {
  if (document.activeElement instanceof HTMLElement) {
    document.activeElement.blur();
  }
};
const DataTableColumnSizing = {
  createHeader(header, table) {
    const clearContainerStyles = (container) => {
      requestAnimationFrame(() => {
        container?.style.removeProperty(
          DATA_TABLE_STRATO_RESIZING_SIZE_CSS_VAR
        );
        document.body.classList.remove(styles.tableColumnResizing);
        if (table.options.enableColumnPinning) {
          container?.querySelectorAll(
            `[style*=${DATA_TABLE_RESIZING_LEFT_POSITION_CSS_VAR}], [style*=${DATA_TABLE_RESIZING_RIGHT_POSITION_CSS_VAR}]`
          ).forEach((elem) => {
            if (elem instanceof HTMLElement) {
              const style = elem.style;
              style.removeProperty(DATA_TABLE_RESIZING_LEFT_POSITION_CSS_VAR);
              style.removeProperty(
                DATA_TABLE_RESIZING_RIGHT_POSITION_CSS_VAR
              );
            }
          });
        }
      });
    };
    const addPinnedColumnsPositionsStyles = (pinnedPosition, container, colId, newPosition) => {
      const resizingPositionVariable = pinnedPosition === "left" ? DATA_TABLE_RESIZING_LEFT_POSITION_CSS_VAR : DATA_TABLE_RESIZING_RIGHT_POSITION_CSS_VAR;
      const targetColumn = container?.querySelector(
        `[role=columnheader][data-headerid=${CSS.escape(colId)}]`
      );
      targetColumn?.style.setProperty(
        resizingPositionVariable,
        `${newPosition}px`
      );
      const targetColumnParentId = table.getColumn(colId)?.parent?.id;
      if (targetColumnParentId) {
        const pinnedState = table.getState().columnPinning;
        const pinnedColumns = pinnedPosition === "left" ? pinnedState.left : pinnedState.right;
        const pinnedChildren = pinnedColumns?.filter(
          (id) => id.includes(targetColumnParentId)
        );
        const edgeChildId = pinnedPosition === "left" ? pinnedChildren?.at(0) : pinnedChildren?.at(-1);
        if (colId === edgeChildId) {
          const parentNode = container?.querySelector(
            `[data-headerid*=${pinnedPosition}][data-headerid*=${CSS.escape(
              targetColumnParentId
            )}]`
          );
          parentNode?.style.setProperty(
            resizingPositionVariable,
            `${newPosition}px`
          );
        }
      }
      const allColumnCells = container?.querySelectorAll(
        `[role=cell][data-columnid=${CSS.escape(colId)}]`
      );
      if (allColumnCells?.length) {
        Array.from(allColumnCells).forEach((cell) => {
          cell.style.setProperty(resizingPositionVariable, `${newPosition}px`);
        });
      }
      const pinnedShadowNodes = container?.querySelectorAll(`[data-pinned-position=${pinnedPosition}]`);
      if (pinnedShadowNodes?.length) {
        Array.from(pinnedShadowNodes).forEach((node) => {
          node.style.setProperty(resizingPositionVariable, `${newPosition}px`);
        });
      }
    };
    header.handleResize = (event) => {
      const column = table.getColumn(header.column.id);
      const canResize = column?.getCanResize();
      const pinnedPosition = header.getIsPinned();
      const isPinnedLeft = pinnedPosition === "left";
      const isPinnedRight = pinnedPosition === "right";
      if (!column || !canResize) {
        return;
      }
      event.persist();
      if (event.detail === 2) {
        return;
      }
      const target = event.target;
      if (target instanceof HTMLElement === false) {
        return;
      }
      const pinnedColumnsWidths = { left: {}, right: {} };
      let columnIndex;
      if (pinnedPosition) {
        const pinnedState = table.getState().columnPinning;
        const headerId = target.parentElement?.dataset.headerid;
        columnIndex = headerId ? pinnedState[pinnedPosition]?.indexOf(headerId) : void 0;
        if (columnIndex !== void 0 && pinnedState[pinnedPosition]?.length) {
          for (const element of pinnedState[pinnedPosition]) {
            pinnedColumnsWidths[pinnedPosition][element] = table.getColumnOrHeaderWidth(element);
          }
        }
      }
      const startSize = getHeaderCellWidth(target);
      if (startSize === void 0) {
        return;
      }
      const columnSizingStart = [[column.id, startSize]];
      const newColumnSizing = {};
      const clientX = getClientX(event.nativeEvent) ?? null;
      const {
        minWidth = DATA_TABLE_DEFAULT_MIN_COLUMN_WIDTH,
        maxWidth = Number.MAX_SAFE_INTEGER
      } = column.columnDef;
      const size = `${Math.min(Math.max(startSize, minWidth), maxWidth)}px`;
      const container = target.closest(
        ".strato-table-virtualization-container"
      );
      container?.style.setProperty(
        DATA_TABLE_STRATO_RESIZING_SIZE_CSS_VAR,
        size
      );
      document.body.classList.add(styles.tableColumnResizing);
      const updateOffset = (eventType, clientXPos) => {
        if (typeof clientXPos !== "number") {
          return;
        }
        const old = table.getState().columnSizingInfo;
        const deltaOffset = clientXPos - (old?.startOffset ?? 0);
        const deltaOffsetRatio = deltaOffset / (old?.startSize ?? 0);
        const deltaPercentage = isPinnedRight ? deltaOffsetRatio : Math.max(deltaOffsetRatio, -0.9999999);
        columnSizingStart.forEach(([columnId, headerSize]) => {
          newColumnSizing[columnId] = Math.round(
            Math.max(
              isPinnedRight ? headerSize - headerSize * deltaPercentage : headerSize + headerSize * deltaPercentage,
              0
            ) * 100
          ) / 100;
        });
        const {
          minWidth: minWidth2 = DATA_TABLE_DEFAULT_MIN_COLUMN_WIDTH,
          maxWidth: maxWidth2 = Number.MAX_SAFE_INTEGER
        } = column.columnDef;
        const size2 = `${Math.min(
          Math.max(
            isPinnedRight ? startSize - deltaOffset : startSize + deltaOffset,
            minWidth2
          ),
          maxWidth2
        )}px`;
        container?.style.setProperty(
          DATA_TABLE_STRATO_RESIZING_SIZE_CSS_VAR,
          size2
        );
        updatePinnedColumnsPositions(deltaOffset);
        if (eventType === "end") {
          table.setColumnSizing((old2) => ({
            ...old2,
            ...newColumnSizing
          }));
        }
      };
      const updatePinnedColumnsPositions = (deltaOffset) => {
        if (columnIndex === void 0) {
          return;
        }
        const leftPinnedColumnsWidths = pinnedColumnsWidths.left;
        const rightPinnedColumnsWidths = pinnedColumnsWidths.right;
        if (isPinnedLeft) {
          const values = Object.entries(leftPinnedColumnsWidths);
          const [, currentColumnInitialWidth] = values[columnIndex];
          for (let i = columnIndex + 1; i < values.length; i++) {
            const [colId] = values[i];
            let prevColumnIndex = i - 1;
            let totalWidth = 0;
            while (prevColumnIndex >= 0) {
              const [, prevColWidth] = values[prevColumnIndex];
              totalWidth += prevColWidth;
              prevColumnIndex -= 1;
            }
            const leftPositionThreshold = totalWidth - currentColumnInitialWidth + minWidth;
            const newLeftPosition = Math.max(
              totalWidth + deltaOffset,
              leftPositionThreshold
            );
            addPinnedColumnsPositionsStyles(
              "left",
              container,
              colId,
              newLeftPosition
            );
          }
        }
        if (isPinnedRight) {
          const values = Object.entries(rightPinnedColumnsWidths);
          const [, currentColumnInitialWidth] = values[columnIndex];
          for (let i = columnIndex - 1; i >= 0; i--) {
            const [colId] = values[i];
            let prevColumnIndex = i + 1;
            let totalWidth = 0;
            while (prevColumnIndex < values.length) {
              const [, prevWidth] = values[prevColumnIndex];
              totalWidth += prevWidth;
              prevColumnIndex += 1;
            }
            const rightPositionThreshold = totalWidth - currentColumnInitialWidth + minWidth;
            const newRightPosition = Math.max(
              totalWidth - deltaOffset,
              rightPositionThreshold
            );
            addPinnedColumnsPositionsStyles(
              "right",
              container,
              colId,
              newRightPosition
            );
          }
        }
      };
      const onEnd = (clientXPos) => {
        updateOffset("end", clientXPos);
        clearContainerStyles(container);
        clearResizerFocus();
        table.setColumnSizingInfo((old) => ({
          ...old,
          isResizingColumn: false,
          startOffset: null,
          startSize: null,
          deltaOffset: null,
          deltaPercentage: null,
          columnSizingStart: []
        }));
      };
      const moveHandler = (moveEvent) => {
        const clientXPos = getClientX(moveEvent);
        updateOffset("move", clientXPos);
      };
      const upHandler = (upEvent) => {
        document.removeEventListener("mousemove", moveHandler);
        document.removeEventListener("mouseup", upHandler);
        document.removeEventListener("touchmove", moveHandler);
        document.removeEventListener("touchend", upHandler);
        const clientXPos = getClientX(upEvent);
        onEnd(clientXPos);
      };
      document.addEventListener("mousemove", moveHandler);
      document.addEventListener("mouseup", upHandler);
      document.addEventListener("touchmove", moveHandler);
      document.addEventListener("touchend", upHandler);
      if (table.getHasFractionalTemplate()) {
        const fractionColumns = table.getVisibleLeafColumnsWithoutBuiltin().filter(
          ({ id, columnDef }) => (columnDef.width === void 0 || isFractionWidth(columnDef.width)) && id !== column.id
        );
        table.setColumnSizing((old) => {
          const newColumnSizing2 = { ...old };
          for (const fractionColumn of fractionColumns) {
            newColumnSizing2[fractionColumn.id] = table.getColumnOrHeaderWidth(
              fractionColumn.id
            );
          }
          return newColumnSizing2;
        });
      }
      table.setColumnSizingInfo((old) => ({
        ...old,
        startOffset: clientX,
        startSize,
        deltaOffset: 0,
        deltaPercentage: 0,
        columnSizingStart,
        isResizingColumn: column.id
      }));
    };
    header.handleResizeExpand = (event) => {
      const column = table.getColumn(header.column.id);
      const canResize = column?.getCanResize();
      if (event.type !== "dblclick") {
        return;
      }
      if (!column || !canResize) {
        return;
      }
      event.persist();
      const target = event.target;
      if (target instanceof HTMLElement === false) {
        return;
      }
      const startSize = getHeaderCellWidth(target);
      if (startSize === void 0) {
        return;
      }
      const clientX = event.clientX;
      const columnSizingStart = [[column.id, startSize]];
      const container = target.closest(
        ".strato-table-virtualization-container"
      );
      container?.style.setProperty(
        DATA_TABLE_STRATO_RESIZING_SIZE_CSS_VAR,
        "max-content"
      );
      table.setColumnSizingInfo((old) => ({
        ...old,
        startOffset: clientX,
        startSize,
        deltaOffset: 0,
        deltaPercentage: 0,
        columnSizingStart,
        isResizingColumn: column.id
      }));
      table._queue(() => {
        const target2 = event.target;
        if (!(target2 instanceof HTMLButtonElement)) {
          return;
        }
        const cell = target2.parentElement;
        if (cell) {
          const { width } = cell.getBoundingClientRect();
          table.setColumnSizingInfo((old) => ({
            ...old,
            isResizingColumn: false,
            startOffset: null,
            startSize: null,
            deltaOffset: null,
            deltaPercentage: null,
            columnSizingStart: []
          }));
          table.setColumnSizing((old) => ({
            ...old,
            [column.id]: width
          }));
        }
        requestAnimationFrame(
          () => container?.style.removeProperty(
            DATA_TABLE_STRATO_RESIZING_SIZE_CSS_VAR
          )
        );
      });
    };
    header.handleKeyboardResize = (event) => {
      const column = table.getColumn(header.column.id);
      const canResize = column?.getCanResize();
      const target = event.target;
      if (!column || !canResize || target instanceof HTMLElement === false || event.key !== "ArrowLeft" && event.key !== "ArrowRight") {
        if (event.key === "Escape") {
          clearResizerFocus();
        }
        return;
      }
      event.persist();
      const startSize = getHeaderCellWidth(target);
      if (startSize === void 0) {
        return;
      }
      let offset = 0;
      if (event.key === "ArrowLeft") {
        offset = -COLUMN_RESIZE_KEYBOARD_OFFSET;
      } else if (event.key === "ArrowRight") {
        offset = COLUMN_RESIZE_KEYBOARD_OFFSET;
      }
      if (event.ctrlKey === true) {
        offset *= COLUMN_RESIZE_KEYBOARD_CTRL_MULTIPLIER;
      }
      table.setColumnSizing((old) => ({
        ...old,
        [column.id]: startSize + offset
      }));
    };
  },
  createColumn(column, table) {
    column.getCanResize = () => {
      if (column.columns.length !== 0) {
        return false;
      }
      if (table.options.enableColumnResizing !== true || column.columnDef.enableResizing === false) {
        return false;
      }
      const lastLeafColumn = table.getVisibleLeafColumns().at(-1);
      if (column.id === lastLeafColumn?.id) {
        return false;
      }
      return true;
    };
  }
};
function useColumnSizing(props, options) {
  options.enableColumnResizing = false;
  const configResizable = useDataTableConfigElement("resizable");
  const [columnSizing, setColumnSizing] = useControllableState({
    value: props.columnSizing,
    defaultValue: props.defaultColumnSizing ?? {},
    onChange: props.onColumnSizingChange
  });
  if (configResizable === true || props.resizable === true) {
    options.enableColumnResizing = true;
    options.state = {
      ...options.state,
      columnSizing
    };
    options.onColumnSizingChange = (updater) => {
      const newValue = functionalUpdate(updater, columnSizing);
      setColumnSizing(newValue);
    };
  }
}
export {
  DataTableColumnSizing,
  useColumnSizing
};
