// @dynatrace/strato-components/charts/core/hooks/useDragElement.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useDragElement_exports = {};
__export(useDragElement_exports, {
  useDragElement: () => useDragElement
});
module.exports = __toCommonJS(useDragElement_exports);
var import_react = require("@use-gesture/react");
var import_react2 = require("react");
var import_fitBetween = require("../utils/fitBetween.js");
function useDragElement(toolbarElement, maxBounds, initialPosition) {
  const [isDragging, setIsDragging] = (0, import_react2.useState)(false);
  const [didDrag, setDidDrag] = (0, import_react2.useState)(false);
  const triggerDragRef = (0, import_react2.useRef)(null);
  const [currentPosition, setCurrentPosition] = (0, import_react2.useState)(initialPosition);
  const [lastIdlePosition, setLastIdlePosition] = (0, import_react2.useState)(initialPosition);
  (0, import_react2.useLayoutEffect)(() => {
    setCurrentPosition((prevCurrentPosition) => {
      const right = (0, import_fitBetween.fitBetween)(prevCurrentPosition.right, 0, maxBounds.right);
      const top = (0, import_fitBetween.fitBetween)(prevCurrentPosition.top, 0, maxBounds.top);
      return { right, top };
    });
    setLastIdlePosition((prevIdlePosition) => {
      const right = (0, import_fitBetween.fitBetween)(prevIdlePosition.right, 0, maxBounds.right);
      const top = (0, import_fitBetween.fitBetween)(prevIdlePosition.top, 0, maxBounds.top);
      return { right, top };
    });
  }, [maxBounds]);
  (0, import_react2.useEffect)(() => {
    if (didDrag) {
      return;
    }
    setCurrentPosition(initialPosition);
    setLastIdlePosition(initialPosition);
  }, [initialPosition.top, initialPosition.right]);
  (0, import_react2.useEffect)(() => {
    const { top, right } = currentPosition;
    if (toolbarElement) {
      toolbarElement.style.right = `${right}px`;
      toolbarElement.style.top = `${top}px`;
    }
  }, [toolbarElement, currentPosition]);
  (0, import_react.useGesture)(
    {
      onDrag: (event) => {
        const [movedX, movedY] = event.movement;
        if (event.intentional) {
          const x = lastIdlePosition.right - movedX;
          const y = lastIdlePosition.top + movedY;
          const right = (0, import_fitBetween.fitBetween)(x, 0, maxBounds.right);
          const top = (0, import_fitBetween.fitBetween)(y, 0, maxBounds.top);
          setCurrentPosition({ right, top });
          setIsDragging(true);
          setDidDrag(true);
        }
      },
      onDragEnd: () => {
        setLastIdlePosition(currentPosition);
        setIsDragging(false);
      }
    },
    {
      target: triggerDragRef,
      eventOptions: { passive: false },
      drag: { filterTaps: true, threshold: 8 }
    }
  );
  return { isDragging, triggerDragRef, didDrag, setDidDrag };
}
