// @dynatrace/strato-components/charts/core/hooks/useFontSizeForContainer.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useFontSizeForContainer_exports = {};
__export(useFontSizeForContainer_exports, {
  useFontSizeForContainer: () => useFontSizeForContainer
});
module.exports = __toCommonJS(useFontSizeForContainer_exports);
var import_react = require("react");
var import_constants = require("../constants.js");
var import_measure_font_baseline = require("../utils/measure-font-baseline.js");
const SPACING_RATIO_THRESHOLD = 1.5;
const MAX_HORIZONTAL_WEIGHT = 0.8;
const BASE_HORIZONTAL_WEIGHT = 0.5;
const WEIGHT_ADJUSTMENT_FACTOR = 0.1;
const VERTICAL_CONSTRAINT_FACTOR = 1.1;
function useFontSizeForContainer({
  container,
  text,
  element,
  suffix = "",
  suffixElement
}) {
  const style = element ? getComputedStyle(element) : {};
  const valueBaseline = (0, import_react.useMemo)(
    () => (0, import_measure_font_baseline.measureFontBaseline)({
      text,
      fontStyle: style.fontStyle,
      fontWeight: style.fontWeight,
      fontFamily: style.fontFamily
    }),
    [text, style.fontStyle, style.fontWeight, style.fontFamily]
  );
  const suffixStyle = suffixElement ? getComputedStyle(suffixElement) : {};
  const suffixBaseline = (0, import_react.useMemo)(
    () => (0, import_measure_font_baseline.measureFontBaseline)({
      text: suffix,
      fontStyle: suffixStyle.fontStyle,
      fontWeight: suffixStyle.fontWeight,
      fontFamily: suffixStyle.fontFamily
    }),
    [
      suffix,
      suffixStyle.fontStyle,
      suffixStyle.fontWeight,
      suffixStyle.fontFamily
    ]
  );
  function calculateWeightedFontSize(fontSizeFitHorizontal, fontSizeFitVertical, spacingRatio) {
    const horizontalWeight = Math.min(
      MAX_HORIZONTAL_WEIGHT,
      BASE_HORIZONTAL_WEIGHT + (spacingRatio - SPACING_RATIO_THRESHOLD) * WEIGHT_ADJUSTMENT_FACTOR
    );
    const verticalWeight = 1 - horizontalWeight;
    const weightedAverage = fontSizeFitHorizontal * horizontalWeight + fontSizeFitVertical * verticalWeight;
    return Math.min(
      weightedAverage,
      fontSizeFitVertical * VERTICAL_CONSTRAINT_FACTOR
    );
  }
  function calculateStandardFontSize(fontSizeFitHorizontal, fontSizeFitVertical) {
    return Math.min(fontSizeFitHorizontal, fontSizeFitVertical);
  }
  if (valueBaseline) {
    const suffixWidth = suffixBaseline ? suffixBaseline.width / import_constants.FONT_BASE * import_constants.SUFFIX_RATIO : 0;
    const fontSizeFitHorizontal = container.width / (valueBaseline.width / import_constants.FONT_BASE + suffixWidth);
    const fontSizeFitVertical = container.height / (valueBaseline.height / import_constants.FONT_BASE);
    const spacingRatio = fontSizeFitHorizontal / fontSizeFitVertical;
    let preferredFontSize;
    if (spacingRatio > SPACING_RATIO_THRESHOLD) {
      preferredFontSize = calculateWeightedFontSize(
        fontSizeFitHorizontal,
        fontSizeFitVertical,
        spacingRatio
      );
    } else {
      preferredFontSize = calculateStandardFontSize(
        fontSizeFitHorizontal,
        fontSizeFitVertical
      );
    }
    return isFinite(preferredFontSize) ? preferredFontSize : 0;
  }
  return 0;
}
