// @dynatrace/strato-components/charts/core/hooks/useObserveContainerSize.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useObserveContainerSize_exports = {};
__export(useObserveContainerSize_exports, {
  useObserveContainerSize: () => useObserveContainerSize
});
module.exports = __toCommonJS(useObserveContainerSize_exports);
var import_lodash_es = require("lodash");
var import_react = require("react");
function calculateNodeDimensions(node) {
  const getNodePaddings = (node2) => {
    const coercePadding = (padding2) => (0, import_lodash_es.toNumber)(padding2.match(/^\d+/g)?.join("") ?? 0);
    const computedElStyle = getComputedStyle(node2);
    return {
      top: coercePadding(computedElStyle.paddingTop),
      bottom: coercePadding(computedElStyle.paddingBottom),
      left: coercePadding(computedElStyle.paddingLeft),
      right: coercePadding(computedElStyle.paddingRight)
    };
  };
  const { width, height } = node.getBoundingClientRect();
  const padding = getNodePaddings(node);
  const { height: wrapperHeight } = node.parentElement?.getBoundingClientRect() ?? { height: 0 };
  return {
    width: width - padding.left - padding.right,
    height: height - padding.top - padding.bottom,
    padding,
    wrapperHeight: wrapperHeight - padding.top - padding.bottom
  };
}
const useObserveContainerSize = () => {
  const [dimensions, setDimensions] = (0, import_react.useState)({
    width: 0,
    height: 0,
    padding: { top: 0, bottom: 0, left: 0, right: 0 },
    wrapperHeight: 0
  });
  const observerCleanupRef = (0, import_react.useRef)(void 0);
  const ref = (0, import_react.useCallback)((node) => {
    if (node === null) {
      if (observerCleanupRef.current) {
        observerCleanupRef.current();
      }
      return;
    }
    setDimensions(calculateNodeDimensions(node));
    const observer = new ResizeObserver(() => {
      setDimensions(calculateNodeDimensions(node));
    });
    observerCleanupRef.current = () => observer.unobserve(node);
    observer.observe(node);
  }, []);
  return [dimensions, ref];
};
