// @dynatrace/strato-components/charts/core/hooks/usePointerEvents.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var usePointerEvents_exports = {};
__export(usePointerEvents_exports, {
  usePointerEvents: () => usePointerEvents
});
module.exports = __toCommonJS(usePointerEvents_exports);
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_constants = require("../constants.js");
var import_pointer_events = require("../types/pointer-events.js");
var import_calculate_mouse_positions = require("../utils/calculate-mouse-positions.js");
var import_drag_event_helpers = require("../utils/drag-event-helpers.js");
var import_get_touches_distance = require("../utils/get-touches-distance.js");
const isIntentionalMovement = (startedHoldingAt, position) => position.relative.x > startedHoldingAt.relative.x + 5 || position.relative.x < startedHoldingAt.relative.x - 5 || position.relative.y > startedHoldingAt.relative.y + 5 || position.relative.y < startedHoldingAt.relative.y - 5;
const extractEventModifiers = ({
  altKey,
  ctrlKey,
  metaKey,
  shiftKey
}) => ({
  altKey,
  ctrlKey,
  metaKey,
  shiftKey
});
const usePointerEvents = ({
  onDoubleTap,
  onDragEnd,
  onDragMove,
  onDragStart,
  onMouseClick,
  onMouseDoubleClick,
  onPinchEnd,
  onPinchMove,
  onPinchStart,
  onTap,
  onMouseHover,
  onTouchStart,
  onDragStartWithMiddleButton,
  onDragMoveWithMiddleButton,
  onDragEndWithMiddleButton,
  onSwipeStart,
  onSwipeMove,
  onSwipeEnd,
  eventCancelled
}) => {
  const currentEventRef = (0, import_react.useRef)({
    action: import_pointer_events.PointerEventActions.Idle
  });
  const startedHoldingAtRef = (0, import_react.useRef)(null);
  const startedTouchesAtRef = (0, import_react.useRef)(null);
  const timeoutRef = (0, import_react.useRef)(null);
  const triggerDblClickTimeout = () => {
    timeoutRef.current = setTimeout(() => {
      clearDblClickTimeout();
    }, 200);
  };
  const clearDblClickTimeout = () => {
    if (timeoutRef.current) {
      clearTimeout(timeoutRef.current);
      timeoutRef.current = null;
    }
  };
  const dispatchHoldAndDragStart = ({ action, ...payload }, eventModifiers) => {
    if (action === import_pointer_events.PointerEventActions.MiddleButtonDragStart) {
      onDragStartWithMiddleButton?.({ ...payload, ...eventModifiers });
    } else {
      onDragStart?.({ ...payload, ...eventModifiers });
    }
  };
  const dispatchHoldAndDragMove = ({ action, ...payload }, eventModifiers) => {
    if (action === import_pointer_events.PointerEventActions.MiddleButtonDragMove) {
      onDragMoveWithMiddleButton?.({ ...payload, ...eventModifiers });
    } else {
      onDragMove?.({ ...payload, ...eventModifiers });
    }
  };
  const dispatchHoldAndDragEnd = ({ action, ...payload }, eventModifiers) => {
    if (action === import_pointer_events.PointerEventActions.MiddleButtonDragEnd) {
      onDragEndWithMiddleButton?.({ ...payload, ...eventModifiers });
    } else {
      onDragEnd?.({ ...payload, ...eventModifiers });
    }
  };
  const getGestureAndCallback = (createTouchEvents) => {
    const callbackOnce = createTouchEvents ? onTap : onMouseClick;
    const callbackTwice = createTouchEvents ? onDoubleTap : onMouseDoubleClick;
    const clickedOnceAction = createTouchEvents ? import_pointer_events.PointerEventActions.Tap : import_pointer_events.PointerEventActions.Click;
    const clickedTwiceAction = createTouchEvents ? import_pointer_events.PointerEventActions.DblTap : import_pointer_events.PointerEventActions.DblClick;
    return {
      callbackOnce,
      callbackTwice,
      clickedOnceAction,
      clickedTwiceAction
    };
  };
  function handleAmountOfClicks(mousePosition, createTouchEvents = false, eventModifiers) {
    let gesture;
    const {
      callbackOnce,
      callbackTwice,
      clickedOnceAction,
      clickedTwiceAction
    } = getGestureAndCallback(createTouchEvents);
    if ((0, import_lodash_es.isNull)(timeoutRef.current)) {
      callbackOnce?.({ position: mousePosition, ...eventModifiers });
      triggerDblClickTimeout();
      gesture = {
        action: clickedOnceAction,
        position: mousePosition
      };
    } else {
      clearDblClickTimeout();
      callbackTwice?.({ position: mousePosition, ...eventModifiers });
      gesture = {
        action: clickedTwiceAction,
        position: mousePosition
      };
    }
    return gesture;
  }
  const handleMouseDown = (mouseEvent) => {
    if (mouseEvent.buttons === import_constants.RIGHT_BUTTON_ID) {
      return;
    }
    const mousePosition = (0, import_calculate_mouse_positions.calculateMousePositions)(
      mouseEvent,
      mouseEvent.currentTarget
    );
    if ((0, import_lodash_es.isUndefined)(mousePosition)) {
      return;
    }
    startedHoldingAtRef.current = mousePosition;
  };
  const handleMouseMove = (mouseEvent) => {
    if (mouseEvent.buttons === import_constants.RIGHT_BUTTON_ID) {
      return;
    }
    const position = (0, import_calculate_mouse_positions.calculateMousePositions)(
      mouseEvent,
      mouseEvent.currentTarget
    );
    if ((0, import_lodash_es.isUndefined)(position)) {
      return;
    }
    let gesture;
    if ((0, import_pointer_events.isDragMove)(currentEventRef.current) || (0, import_pointer_events.isMiddleButtonDragMove)(currentEventRef.current)) {
      gesture = (0, import_drag_event_helpers.buildDragMoveEvent)(
        mouseEvent.buttons,
        currentEventRef.current.startHoldingAt,
        currentEventRef.current.currentPosition,
        position
      );
      dispatchHoldAndDragMove(gesture, extractEventModifiers(mouseEvent));
    } else if (startedHoldingAtRef.current && isIntentionalMovement(startedHoldingAtRef.current, position)) {
      const eventToDispatch = (0, import_drag_event_helpers.buildDragStartEvent)(
        mouseEvent.buttons,
        startedHoldingAtRef.current
      );
      dispatchHoldAndDragStart(
        eventToDispatch,
        extractEventModifiers(mouseEvent)
      );
      gesture = (0, import_drag_event_helpers.buildDragMoveEvent)(
        mouseEvent.buttons,
        startedHoldingAtRef.current,
        startedHoldingAtRef.current,
        position
      );
    } else {
      gesture = {
        action: import_pointer_events.PointerEventActions.Hover,
        position
      };
      onMouseHover?.({ position, ...extractEventModifiers(mouseEvent) });
    }
    currentEventRef.current = gesture;
  };
  const handleMouseUp = (mouseEvent) => {
    if (mouseEvent.buttons === import_constants.RIGHT_BUTTON_ID) {
      return;
    }
    if (eventCancelled?.current) {
      eventCancelled.current = false;
      startedHoldingAtRef.current = null;
      currentEventRef.current = {
        action: import_pointer_events.PointerEventActions.Idle
      };
      return;
    }
    const mousePosition = (0, import_calculate_mouse_positions.calculateMousePositions)(
      mouseEvent,
      mouseEvent.currentTarget
    );
    if ((0, import_lodash_es.isUndefined)(mousePosition)) {
      return;
    }
    if ((0, import_pointer_events.isDragMove)(currentEventRef.current) || (0, import_pointer_events.isMiddleButtonDragMove)(currentEventRef.current)) {
      const gesture = (0, import_drag_event_helpers.buildDragEndEvent)(
        currentEventRef.current,
        currentEventRef.current.startHoldingAt,
        mousePosition
      );
      dispatchHoldAndDragEnd(gesture, extractEventModifiers(mouseEvent));
      currentEventRef.current = { ...gesture };
    } else {
      currentEventRef.current = handleAmountOfClicks(
        mousePosition,
        false,
        extractEventModifiers(mouseEvent)
      );
    }
    startedHoldingAtRef.current = null;
  };
  const handleTouchStart = (touchEvent) => {
    const { touches } = touchEvent;
    const mousePosition = (0, import_calculate_mouse_positions.calculateMousePositions)(
      touches[0],
      touchEvent.currentTarget ?? touchEvent.nativeEvent.currentTarget
    );
    if ((0, import_lodash_es.isUndefined)(mousePosition)) {
      return;
    }
    onTouchStart?.({
      position: mousePosition,
      ...extractEventModifiers(touchEvent)
    });
    startedHoldingAtRef.current = mousePosition;
    startedTouchesAtRef.current = touches;
  };
  const handleTouchMove = (touchEvent) => {
    const { touches } = touchEvent;
    let gesture = {
      action: import_pointer_events.PointerEventActions.Idle
    };
    const amountOfTouches = touches.length;
    switch (amountOfTouches) {
      case 1: {
        const mousePosition = (0, import_calculate_mouse_positions.calculateMousePositions)(
          touches[0],
          touchEvent.currentTarget ?? touchEvent.nativeEvent.currentTarget
        );
        if ((0, import_lodash_es.isUndefined)(mousePosition)) {
          return;
        }
        if ((0, import_pointer_events.isDragMove)(currentEventRef.current)) {
          const { startHoldingAt, currentPosition: prevPosition } = currentEventRef.current;
          const delta = mousePosition.relative.x - prevPosition.relative.x;
          gesture = {
            action: import_pointer_events.PointerEventActions.DragMove,
            startHoldingAt,
            currentPosition: mousePosition,
            delta
          };
          onDragMove?.({
            startHoldingAt,
            currentPosition: mousePosition,
            delta,
            ...extractEventModifiers(touchEvent)
          });
        } else if (startedHoldingAtRef.current && isIntentionalMovement(startedHoldingAtRef.current, mousePosition)) {
          const delta = mousePosition.relative.x - startedHoldingAtRef.current.relative.x;
          gesture = {
            action: import_pointer_events.PointerEventActions.DragMove,
            startHoldingAt: startedHoldingAtRef.current,
            currentPosition: mousePosition,
            delta
          };
          onDragStart?.({
            startHoldingAt: startedHoldingAtRef.current,
            ...extractEventModifiers(touchEvent)
          });
        }
        break;
      }
      case 2: {
        const distance = (0, import_get_touches_distance.getTouchesDistance)(touches);
        if ((0, import_pointer_events.isPinchGesture)(currentEventRef.current) || (0, import_pointer_events.isPinchStartGesture)(currentEventRef.current)) {
          let delta, initialDistance;
          if ((0, import_pointer_events.isPinchGesture)(currentEventRef.current)) {
            const prevDistance = currentEventRef.current.distance;
            initialDistance = currentEventRef.current.initialDistance;
            delta = distance - prevDistance;
          } else {
            initialDistance = currentEventRef.current.distance;
            delta = distance - initialDistance;
          }
          const changeRatio = distance / initialDistance;
          const payload = {
            initialDistance,
            changeRatio,
            delta,
            distance
          };
          gesture = {
            action: import_pointer_events.PointerEventActions.Pinch,
            ...payload
          };
          onPinchMove?.({ ...payload, ...extractEventModifiers(touchEvent) });
        } else if ((0, import_pointer_events.isSwipeStart)(currentEventRef.current) || (0, import_pointer_events.isSwipeMove)(currentEventRef.current)) {
          const { position } = currentEventRef.current;
          const deltaXFingerZero = {
            deltaX: touches[0].clientX - position[0].clientX,
            deltaY: touches[0].clientY - position[0].clientY,
            id: touches[0].identifier
          };
          const deltaXFingerOne = {
            deltaX: touches[1].clientX - position[1].clientX,
            deltaY: touches[1].clientY - position[1].clientY,
            id: touches[1].identifier
          };
          const fingerZeroMovedMoreX = Math.abs(deltaXFingerZero.deltaX) > Math.abs(deltaXFingerOne.deltaX);
          const { deltaX } = fingerZeroMovedMoreX ? deltaXFingerZero : deltaXFingerOne;
          const fingerZeroMovedMoreY = Math.abs(deltaXFingerZero.deltaY) > Math.abs(deltaXFingerOne.deltaY);
          const { deltaY } = fingerZeroMovedMoreY ? deltaXFingerZero : deltaXFingerOne;
          let delta;
          let direction;
          if (Math.abs(deltaY) > Math.abs(deltaX)) {
            delta = deltaY;
            direction = deltaY >= 0 ? "down" : "up";
          } else {
            delta = deltaX;
            direction = deltaX >= 0 ? "right" : "left";
          }
          const payload = {
            position: touchEvent.touches,
            delta,
            direction,
            deltaY,
            deltaX
          };
          gesture = {
            action: import_pointer_events.PointerEventActions.SwipeMove,
            ...payload
          };
          onSwipeMove?.({ ...payload, ...extractEventModifiers(touchEvent) });
        } else {
          if (startedTouchesAtRef.current) {
            const initialDistanceBetweenFingers = (0, import_get_touches_distance.getTouchesDistance)(
              startedTouchesAtRef.current
            );
            const currentDistanceBetweenFingers = (0, import_get_touches_distance.getTouchesDistance)(touches);
            const delta = Math.abs(
              initialDistanceBetweenFingers - currentDistanceBetweenFingers
            );
            const pinchThreshold = initialDistanceBetweenFingers * import_constants.PINCH_RELATIVE_THRESHOLD;
            if (delta > pinchThreshold) {
              gesture = {
                action: import_pointer_events.PointerEventActions.PinchStart,
                distance
              };
              onPinchStart?.({
                distance,
                ...extractEventModifiers(touchEvent)
              });
            } else {
              const deltaXFingerZero = startedTouchesAtRef.current[0].clientX - touches[0].clientX;
              const deltaYFingerZero = startedTouchesAtRef.current[0].clientY - touches[0].clientY;
              const delta2 = Math.max(
                Math.abs(deltaXFingerZero),
                Math.abs(deltaYFingerZero)
              );
              const swipeThreshold = initialDistanceBetweenFingers * import_constants.SWIPE_RELATIVE_THRESHOLD;
              if (delta2 > swipeThreshold) {
                gesture = {
                  action: import_pointer_events.PointerEventActions.SwipeStart,
                  position: touches
                };
                onSwipeStart?.({
                  position: touches,
                  ...extractEventModifiers(touchEvent)
                });
              }
            }
          }
        }
        break;
      }
    }
    currentEventRef.current = { ...gesture };
  };
  const handleTouchEnd = (touchEvent) => {
    if (touchEvent.cancelable) {
      touchEvent.preventDefault();
    }
    let gesture = {
      action: import_pointer_events.PointerEventActions.Idle
    };
    const { changedTouches } = touchEvent;
    const amountOfTouches = changedTouches.length;
    switch (amountOfTouches) {
      case 1: {
        const mousePosition = (0, import_calculate_mouse_positions.calculateMousePositions)(
          changedTouches[0],
          touchEvent.currentTarget ?? touchEvent.nativeEvent.currentTarget
        );
        if ((0, import_lodash_es.isUndefined)(mousePosition)) {
          return;
        }
        if ((0, import_pointer_events.isDragMove)(currentEventRef.current)) {
          const { startHoldingAt } = currentEventRef.current;
          gesture = {
            action: import_pointer_events.PointerEventActions.DragEnd,
            startHoldingAt,
            releasedAt: mousePosition
          };
          onDragEnd?.({
            startHoldingAt,
            releasedAt: mousePosition,
            ...extractEventModifiers(touchEvent)
          });
        } else if (startedTouchesAtRef.current?.length === amountOfTouches && startedTouchesAtRef.current[0].identifier === changedTouches[0].identifier) {
          gesture = handleAmountOfClicks(
            mousePosition,
            true,
            extractEventModifiers(touchEvent)
          );
        }
        startedHoldingAtRef.current = null;
        break;
      }
      case 2: {
        if ((0, import_pointer_events.isPinchGesture)(currentEventRef.current)) {
          const { action, ...rest } = currentEventRef.current;
          gesture = {
            ...rest,
            action: import_pointer_events.PointerEventActions.PinchEnd
          };
          onPinchEnd?.({ ...rest, ...extractEventModifiers(touchEvent) });
        } else if ((0, import_pointer_events.isSwipeMove)(currentEventRef.current)) {
          const { action, ...rest } = currentEventRef.current;
          gesture = {
            ...rest,
            action: import_pointer_events.PointerEventActions.SwipeEnd
          };
          onSwipeEnd?.({ ...rest, ...extractEventModifiers(touchEvent) });
        } else {
          gesture = {
            action: import_pointer_events.PointerEventActions.Idle
          };
        }
        startedHoldingAtRef.current = null;
        break;
      }
    }
    currentEventRef.current = { ...gesture };
  };
  return {
    handleMouseDown,
    handleMouseMove,
    handleMouseUp,
    handleTouchStart,
    handleTouchMove,
    handleTouchEnd
  };
};
