// @dynatrace/strato-components/esm/content/chip/BaseChip.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import {
  Children,
  forwardRef,
  Fragment,
  useMemo
} from "react";
import { isElement } from "react-is";
import * as styles from "./BaseChip.sty.js";
import { ChipPrefix, ChipSuffix } from "./ChipContent.js";
import { ChipDeleteButton } from "./ChipDeleteButton.js";
import { ChipKey } from "./ChipKey.js";
import { ChipContext } from "./contexts/ChipContext.js";
import { useChipContext } from "./contexts/useChipContext.js";
import { useAriaLabelingProps } from "../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useId } from "../../core/hooks/useId.js";
import { useSafeForwardProps } from "../../core/hooks/useSafeForwardProps.js";
import { focusRingStatic } from "../../core/styles/focusRingStatic.sty.js";
import { Flex } from "../../layouts/flex/Flex.js";
import { fieldTokens as fieldColorsCSS } from "../../styles/field.sty.js";
import { textStyle as textStyleCSS } from "../../styles/textStyle.sty.js";
import { Text } from "../../typography/text/Text.js";
const COMPONENT_NAME = "Chip";
const getChipSlots = (node) => {
  const slots = {
    deleteButton: null,
    prefixIcons: [],
    suffixIcons: [],
    value: [],
    key: []
  };
  for (const child of Children.toArray(node)) {
    if (isElement(child)) {
      switch (child.type) {
        case ChipPrefix:
          slots.prefixIcons.push(child);
          break;
        case ChipSuffix:
          slots.suffixIcons.push(child);
          break;
        case ChipKey:
          slots.key.push(child);
          break;
        case ChipDeleteButton:
          slots.deleteButton = child;
          break;
        default:
          slots.value.push(child);
      }
    } else {
      slots.value.push(child);
    }
  }
  return slots;
};
const RootBaseChip = /* @__PURE__ */ forwardRef(
  (props, forwardedRef) => {
    const {
      children,
      as = "div",
      variant = "emphasized",
      color = "neutral",
      size = "default",
      disabled = false,
      maxWidth,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "data-dt-component": dataDtComponent = COMPONENT_NAME,
      ...remainingProps
    } = props;
    const ariaLabelingProps = useAriaLabelingProps(remainingProps);
    const forwardProps = useSafeForwardProps(remainingProps, ariaLabelingProps);
    const [behavioralTrackingProps] = useBehavioralTrackingProps(remainingProps);
    const { prefixIcons, suffixIcons, deleteButton, key, value } = useMemo(
      () => getChipSlots(children),
      [children]
    );
    const valueId = useId();
    const hasKey = key.length > 0;
    return /* @__PURE__ */ jsx(
      Flex,
      {
        as: "div",
        ref: forwardedRef,
        gap: 2,
        className: clsx(
          styles.wrapper,
          disabled && styles.disabledWrapper,
          fieldColorsCSS({ variant, color, disabled }),
          consumerClassName
        ),
        style: {
          maxWidth: maxWidth ?? 250,
          ...consumerStyle
        },
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        ...behavioralTrackingProps,
        "data-dt-component": dataDtComponent,
        children: /* @__PURE__ */ jsxs(
          ChipContext.Provider,
          {
            value: { disabled, variant, color, size, valueId, hasKey },
            children: [
              /* @__PURE__ */ jsxs(
                Text,
                {
                  as,
                  textStyle: "base",
                  className: clsx(
                    "strato-surface",
                    focusRingStatic({ color, disabled, when: "focus-visible" }),
                    (as === "a" || as === "button") && textStyleCSS({ textStyle: "base-emphasized" }),
                    fieldColorsCSS({ variant, color, disabled }),
                    styles.container({
                      size,
                      disabled,
                      color,
                      variant,
                      hasDeleteButton: deleteButton !== null
                    })
                  ),
                  size,
                  disabled,
                  variant,
                  color,
                  ...ariaLabelingProps,
                  ...forwardProps,
                  children: [
                    prefixIcons.map((node, index) => /* @__PURE__ */ jsx(Fragment, { children: node }, `prefix-node-${index}`)),
                    /* @__PURE__ */ jsx(TextContainer, { keys: key, values: value }),
                    suffixIcons.map((node, index) => /* @__PURE__ */ jsx(Fragment, { children: node }, `suffix-node-${index}`))
                  ]
                }
              ),
              deleteButton
            ]
          }
        )
      }
    );
  }
);
function TextContainer({
  keys,
  values
}) {
  const { valueId: id } = useChipContext();
  const hasValue = values.length > 0;
  const hasKey = keys.length > 0;
  if (!(hasKey || hasValue)) {
    return null;
  }
  const separator = hasKey && hasValue ? /* @__PURE__ */ jsx("span", { children: ":" }) : null;
  return /* @__PURE__ */ jsxs("div", { className: styles.textContainer, children: [
    keys.map((node, index) => /* @__PURE__ */ jsx(Fragment, { children: node }, `key-node-${index}`)),
    separator,
    values.map((node, index) => /* @__PURE__ */ jsx(Flex, { id, className: styles.value, children: node }, `value-node-${index}`))
  ] });
}
const BaseChip = Object.assign(RootBaseChip, {
  Prefix: ChipPrefix,
  Suffix: ChipSuffix,
  Key: ChipKey,
  DeleteButton: ChipDeleteButton
});
BaseChip.displayName = COMPONENT_NAME;
export {
  BaseChip
};
