// @dynatrace/strato-components/esm/content/chip/ChipDeleteButton.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import { forwardRef } from "react";
import { useIntl } from "react-intl";
import { XmarkIcon } from "@dynatrace/strato-icons";
import * as styles from "./ChipDeleteButton.sty.js";
import { useChipContext } from "./contexts/useChipContext.js";
import { Button } from "../../buttons/button/Button.js";
const ChipDeleteButton = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const intl = useIntl();
  const deleteChip = intl.formatMessage({
    defaultMessage: "Delete chip",
    id: "JfsC0aOiAl7G3esV",
    description: "Aria label used to describe the delete chip button."
  });
  const {
    disabled: propDisabled,
    "aria-label": ariaLabelProp = deleteChip,
    "aria-describedby": ariaDescribedByProp,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    ...remainingProps
  } = props;
  const {
    disabled,
    variant,
    color,
    size = "default",
    valueId
  } = useChipContext();
  return /* @__PURE__ */ jsx(
    Button,
    {
      ref: forwardedRef,
      "data-testid": dataTestId,
      variant,
      color,
      disabled: propDisabled ?? disabled,
      "aria-label": ariaLabelProp,
      "aria-describedby": ariaDescribedByProp || valueId,
      className: clsx(styles.base, styles.size[size], consumerClassName, {
        [styles.neutralEmphasizedOverride]: color === "neutral" && variant === "emphasized",
        [styles.disabledOverride]: propDisabled ?? disabled
      }),
      style: consumerStyle,
      ...remainingProps,
      children: /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(XmarkIcon, { size: size === "default" ? size : "small" }) })
    }
  );
});
ChipDeleteButton.displayName = "Chip.DeleteButton";
export {
  ChipDeleteButton
};
