// @dynatrace/strato-components/esm/content/chip/ChipContent.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import { forwardRef } from "react";
import * as styles from "./ChipContent.sty.js";
import { useChipContext } from "./contexts/useChipContext.js";
const ChipContent = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const {
    children,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    $isPrefixWithKey,
    ...remainingProps
  } = props;
  const { size } = useChipContext();
  return /* @__PURE__ */ jsx(
    "div",
    {
      ref: forwardedRef,
      className: clsx(
        styles.content({
          size
        }),
        consumerClassName
      ),
      style: consumerStyle,
      "data-testid": dataTestId,
      ...remainingProps,
      children
    }
  );
});
const ChipPrefix = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const {
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    ...restProps
  } = props;
  const { hasKey } = useChipContext();
  return /* @__PURE__ */ jsx(
    ChipContent,
    {
      "data-testid": dataTestId,
      ref: forwardedRef,
      className: clsx(
        {
          [styles.prefix]: hasKey
        },
        consumerClassName
      ),
      style: consumerStyle,
      $isPrefixWithKey: hasKey,
      ...restProps
    }
  );
});
const ChipSuffix = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const {
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    ...restProps
  } = props;
  return /* @__PURE__ */ jsx(
    ChipContent,
    {
      "data-testid": dataTestId,
      ref: forwardedRef,
      className: consumerClassName,
      style: consumerStyle,
      ...restProps
    }
  );
});
ChipPrefix.displayName = "Chip.Prefix";
ChipSuffix.displayName = "Chip.Suffix";
export {
  ChipContent,
  ChipPrefix,
  ChipSuffix
};
