// @dynatrace/strato-components/esm/content/chip/GroupedChip.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import {
  forwardRef,
  useEffect,
  useRef
} from "react";
import { BaseChip } from "./BaseChip.js";
import * as styles from "./GroupedChip.sty.js";
import { useContainerSize } from "../../core/hooks/useContainerSize.js";
import { useId } from "../../core/hooks/useId.js";
import { useMergeRefs } from "../../core/hooks/useMergeRefs.js";
import { useChipGroupContext } from "../chip-group/hooks/useChipGroupContext.js";
import { useInternalChipGroupContext } from "../chip-group/hooks/useInternalChipGroupContext.js";
import { ChipGroupActions } from "../chip-group/state/chip-group-state-reducer.js";
const GroupedChip = /* @__PURE__ */ forwardRef(
  (props, forwardedRef) => {
    const {
      children,
      size: propSize,
      disabled: propDisabled,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...remainingProps
    } = props;
    const uniqueId = useId();
    const chipRef = useRef(null);
    const { width } = useContainerSize(chipRef);
    const rootRef = useMergeRefs([chipRef, forwardedRef]);
    const { visibleKeys } = useChipGroupContext();
    const internalChipGroupContext = useInternalChipGroupContext();
    const {
      disabled: chipGroupDisabled,
      size: chipGroupSize,
      dispatch
    } = internalChipGroupContext || {};
    const disabled = propDisabled ?? chipGroupDisabled ?? false;
    const size = propSize ?? chipGroupSize ?? "default";
    useEffect(() => {
      dispatch?.({
        type: ChipGroupActions.ADDCHIP,
        payload: { uniqueId, width }
      });
      return () => {
        dispatch?.({
          type: ChipGroupActions.REMOVECHIP,
          payload: {
            uniqueId
          }
        });
      };
    }, []);
    useEffect(() => {
      dispatch?.({
        type: ChipGroupActions.UPDATECHIP,
        payload: { uniqueId, width: width === 0 ? void 0 : width }
      });
    }, [dispatch, uniqueId, width]);
    return /* @__PURE__ */ jsx(
      "div",
      {
        ref: rootRef,
        className: clsx("chip", styles.grouped, {
          [styles.first]: visibleKeys.includes(uniqueId) && visibleKeys.length === 1,
          [styles.hidden]: !visibleKeys.includes(uniqueId)
        }),
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        children: /* @__PURE__ */ jsx(
          BaseChip,
          {
            ...remainingProps,
            disabled,
            size,
            children
          }
        )
      }
    );
  }
);
export {
  GroupedChip
};
