// @dynatrace/strato-components/content/chip-group/state/chip-group-state-reducer.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var chip_group_state_reducer_exports = {};
__export(chip_group_state_reducer_exports, {
  ChipGroupActions: () => ChipGroupActions,
  chipGroupReducer: () => chipGroupReducer
});
module.exports = __toCommonJS(chip_group_state_reducer_exports);
var ChipGroupActions = /* @__PURE__ */ ((ChipGroupActions2) => {
  ChipGroupActions2["ADDCHIP"] = "add";
  ChipGroupActions2["UPDATECHIP"] = "update";
  ChipGroupActions2["REMOVECHIP"] = "remove";
  ChipGroupActions2["ADDCONTROL"] = "add-control";
  ChipGroupActions2["UPDATECONTROL"] = "update-control";
  ChipGroupActions2["REMOVECONTROL"] = "remove-control";
  return ChipGroupActions2;
})(ChipGroupActions || {});
function chipGroupReducer(state, action) {
  const { type, payload } = action;
  const newState = { ...state };
  if (type === "add" /* ADDCHIP */) {
    const { width } = payload;
    newState.chips = { ...newState.chips, [payload.uniqueId]: { width } };
    return newState;
  }
  if (type === "update" /* UPDATECHIP */) {
    const { uniqueId: _, ...chipState } = payload;
    newState.chips = {
      ...newState.chips,
      [payload.uniqueId]: {
        ...newState.chips[payload.uniqueId],
        ...chipState
      }
    };
    return newState;
  }
  if (type === "remove" /* REMOVECHIP */) {
    delete newState.chips[payload.uniqueId];
    return newState;
  }
  if (type === "add-control" /* ADDCONTROL */) {
    const { width } = payload;
    newState.control = { width, show: "auto" };
    return newState;
  }
  if (type === "update-control" /* UPDATECONTROL */) {
    const { width, show } = payload;
    newState.control = { width, show };
    return newState;
  }
  if (type === "remove-control" /* REMOVECONTROL */) {
    newState.control = void 0;
    return newState;
  }
  return state;
}
