// @dynatrace/strato-components/esm/filters/segment-selector/parts/segmentsPresets/SegmentsPresetsList.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { useQuery } from "@tanstack/react-query";
import clsx from "clsx";
import { AnimatePresence, motion } from "framer-motion";
import { useState } from "react";
import { FormattedMessage } from "react-intl";
import Spacings from "@dynatrace/strato-design-tokens/spacings";
import { getAppName } from "@dynatrace-sdk/app-environment";
import { getSegmentsPresets } from "@dynatrace-sdk/segments-presets";
import { isSelectedPreset } from "./isSelectedPreset.js";
import {
  expandAndCollapseAnimationVariants,
  SegmentsPreset
} from "./SegmentsPreset.js";
import * as styles from "./SegmentsPresets.sty.js";
import { Divider } from "../../../../layouts/divider/Divider.js";
import { Flex } from "../../../../layouts/flex/Flex.js";
import { Text } from "../../../../typography/text/Text.js";
import { segmentsClient } from "../../contexts/SegmentsCache/SegmentsCache.js";
import { useSegments } from "../../hooks/useSegments.js";
import * as animationHelperStyles from "../Overlay.sty.js";
import { ShowMorePanel } from "../ShowMorePanel/ShowMorePanel.js";
const PIN_LIMIT = 10;
const SegmentsPresetList = ({
  showDivider
}) => {
  const { segments } = useSegments();
  const { data, isLoading } = useQuery(
    {
      queryKey: ["getSegmentsPresets"],
      queryFn: getSegmentsPresets,
      notifyOnChangeProps: ["dataUpdatedAt"]
    },
    segmentsClient.queryClient
  );
  const [isAnimating, setIsAnimating] = useState(false);
  if (!data || data.pinnedPresets.length === 0 && data.recentPresets.length === 0 || isLoading) {
    return null;
  }
  const pinLimitReached = data.pinnedPresets.length >= PIN_LIMIT;
  const recentPresets = data.recentPresets?.map((preset) => /* @__PURE__ */ jsx(
    "li",
    {
      className: styles.listItem,
      "data-dt-name": "Recent",
      children: /* @__PURE__ */ jsx(
        SegmentsPreset,
        {
          uiPreset: preset,
          id: preset.id,
          pinLimitReached,
          isPinned: data.pinnedPresets.some(
            (pinnedPreset) => preset.id === pinnedPreset.id
          ),
          isSelected: isSelectedPreset(segments, preset.segments)
        }
      )
    },
    `recent-${preset.id}`
  ));
  const primaryContent = recentPresets.length > 3 ? recentPresets.slice(0, 2) : recentPresets;
  const extendedContent = recentPresets.length > 3 ? recentPresets.slice(2) : null;
  const showRecent = data.recentPresets.length > 0;
  const showPinned = data.pinnedPresets.length > 0;
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    showDivider && /* @__PURE__ */ jsx(Divider, {}),
    /* @__PURE__ */ jsxs(
      Flex,
      {
        flexDirection: "column",
        width: "100%",
        px: 0,
        pb: 4,
        gap: 0,
        pt: showDivider ? 12 : 0,
        className: clsx({
          [animationHelperStyles.isAnimating]: isAnimating
        }),
        children: [
          showRecent && /* @__PURE__ */ jsxs(Flex, { flexDirection: "column", gap: 0, width: "100%", children: [
            /* @__PURE__ */ jsx(
              Text,
              {
                textStyle: "small-emphasized",
                className: styles.sectionHeadline,
                children: /* @__PURE__ */ jsx(
                  FormattedMessage,
                  {
                    id: "QZdQoUnAWKJYAyK2",
                    defaultMessage: "Recently used",
                    description: "Headline to group recently used segments presets."
                  }
                )
              }
            ),
            /* @__PURE__ */ jsx(AnimatePresence, { initial: false, children: /* @__PURE__ */ jsxs(ShowMorePanel, { children: [
              /* @__PURE__ */ jsx(ShowMorePanel.PrimaryContent, { children: primaryContent }),
              extendedContent && /* @__PURE__ */ jsx(ShowMorePanel.ExtendedContent, { children: extendedContent })
            ] }) }, "recent-presets")
          ] }),
          /* @__PURE__ */ jsx(AnimatePresence, { initial: false, children: showPinned && /* @__PURE__ */ jsxs(
            Flex,
            {
              as: motion.div,
              flexDirection: "column",
              gap: 4,
              variants: expandAndCollapseAnimationVariants,
              initial: "exit",
              animate: "enter",
              exit: "exit",
              onAnimationStart: () => setIsAnimating(true),
              onAnimationComplete: () => setIsAnimating(false),
              children: [
                showRecent && showPinned && /* @__PURE__ */ jsx(Divider, {}),
                /* @__PURE__ */ jsx(
                  Text,
                  {
                    textStyle: "small-emphasized",
                    className: styles.sectionHeadline,
                    style: { paddingTop: Spacings.Size12 },
                    children: /* @__PURE__ */ jsx(
                      FormattedMessage,
                      {
                        id: "568KZGFcxrq95nEC",
                        defaultMessage: "Pinned in {appName}",
                        description: "Headline to group recently used segments presets.",
                        values: { appName: getAppName() }
                      },
                      "pinned-header-text"
                    )
                  },
                  "pinned-header"
                ),
                /* @__PURE__ */ jsx("ul", { className: styles.list, children: /* @__PURE__ */ jsx(AnimatePresence, { initial: false, children: data.pinnedPresets.map((preset) => /* @__PURE__ */ jsx(
                  "li",
                  {
                    className: styles.listItem,
                    "data-dt-name": "Pinned",
                    children: /* @__PURE__ */ jsx(
                      SegmentsPreset,
                      {
                        uiPreset: preset,
                        id: preset.id,
                        pinLimitReached,
                        isPinned: true,
                        isSelected: isSelectedPreset(segments, preset.segments)
                      }
                    )
                  },
                  `pinned-${preset.id}`
                )) }, "pinned-presets-list") })
              ]
            }
          ) }, "pinned-presets")
        ]
      }
    )
  ] });
};
export {
  SegmentsPresetList
};
