// @dynatrace/strato-components/esm/filters/segment-selector/parts/segmentsPresets/SegmentsPreset.js
import { jsx, jsxs } from "react/jsx-runtime";
import { useMutation, useQuery } from "@tanstack/react-query";
import clsx from "clsx";
import { motion, useReducedMotion } from "framer-motion";
import { Fragment, useContext, useRef, useState } from "react";
import { FormattedMessage, useIntl } from "react-intl";
import AnimationsJS from "@dynatrace/strato-design-tokens/animations-js";
import { PinIcon, PinOffIcon } from "@dynatrace/strato-icons";
import { getAppName } from "@dynatrace-sdk/app-environment";
import {
  pinSegmentsPreset,
  unpinSegmentsPreset
} from "@dynatrace-sdk/segments-presets";
import * as styles from "./SegmentsPresets.sty.js";
import { Button } from "../../../../buttons/button/Button.js";
import { useOverlayStacking } from "../../../../core/hooks/useOverlayStacking.js";
import { expandCollapseEasing } from "../../../../core/utils/framer-motion-easings.js";
import { Flex } from "../../../../layouts/flex/Flex.js";
import { InputGroup } from "../../../../layouts/input-group/InputGroup.js";
import { Tooltip } from "../../../../overlays/tooltip/Tooltip.js";
import { textStyle as textStyleCSS } from "../../../../styles/textStyle.sty.js";
import { Text } from "../../../../typography/text/Text.js";
import {
  handleFailedOptimisticMutation,
  handleSettledOptimisticMutation,
  runOptimisticMutation
} from "../../../helper/queryHelper.js";
import {
  segmentsClient,
  useSegmentsCache
} from "../../contexts/SegmentsCache/SegmentsCache.js";
import { useSegmentsUIContext } from "../../contexts/SegmentsUI/SegmentsUIContext.js";
import { UISegmentsDispatchContext } from "../../contexts/SegmentsUI/SegmentsUIDispatchContext.js";
import { areUiSegmentsValid } from "../../contexts/SegmentsUI/segmentValidationHelper.js";
import { useSegments } from "../../hooks/useSegments.js";
import { flattenQueryResultToValues, getReadableValue } from "../../utils.js";
import * as animationHelperStyles from "../Overlay.sty.js";
const expandAndCollapseAnimationVariants = {
  enter: {
    height: "auto",
    opacity: 1,
    translateY: 0,
    transition: {
      duration: AnimationsJS.ExpandCollapse.Duration.Base,
      ease: expandCollapseEasing
    }
  },
  exit: {
    height: 0,
    opacity: 0,
    translateY: -10,
    transition: {
      duration: AnimationsJS.ExpandCollapse.Duration.Base,
      ease: expandCollapseEasing
    }
  }
};
const SegmentsPresetText = ({
  segment,
  isFirst
}) => {
  const andText = /* @__PURE__ */ jsxs(Text, { textStyle: "base", className: styles.and, children: [
    "\xA0",
    /* @__PURE__ */ jsx(
      FormattedMessage,
      {
        defaultMessage: "and",
        description: "Text that is separating between multiple Segments in one preset.",
        id: "GeQL1Mbr4RzmlFJk"
      }
    ),
    "\xA0"
  ] });
  const connectingString = isFirst ? "" : andText;
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    connectingString,
    /* @__PURE__ */ jsx(SegmentText, { segment })
  ] }, segment.id);
};
function usePinningTooltipText(pinLimitReached, isPinned) {
  const intl = useIntl();
  if (pinLimitReached && !isPinned) {
    return intl.formatMessage({
      defaultMessage: "You can't pin more than 10 items.",
      id: "Y/6wwuz1ueL+7V2j",
      description: "Tooltip text for pinning limit."
    });
  }
  if (!isPinned) {
    return intl.formatMessage(
      {
        defaultMessage: "Pin in {appName}",
        id: "RPdRdtDJumoOO3os",
        description: "Tooltip text for pin button."
      },
      { appName: getAppName() }
    );
  }
  return intl.formatMessage({
    defaultMessage: "Unpin",
    id: "uCbwlgMp2BNGzNro",
    description: "Text for the unpin button of a segment preset."
  });
}
function updateCachePinMutation(old, id, segments) {
  const pinnedPresets = old.pinnedPresets;
  pinnedPresets.unshift({ id, segments });
  return {
    ...old,
    pinnedPresets
  };
}
function updateCacheUnpinMutation(old, id) {
  const pinnedPresets = old.pinnedPresets;
  const pinnedPresetIndex = pinnedPresets.findIndex(
    (preset) => preset.id === id
  );
  if (pinnedPresetIndex === -1) {
    return old;
  }
  pinnedPresets.splice(pinnedPresetIndex, 1);
  return {
    recentPresets: old.recentPresets,
    pinnedPresets
  };
}
const PresetTooltipText = ({ segments }) => {
  return /* @__PURE__ */ jsx(Flex, { flexDirection: "column", gap: 4, children: segments.map((segment) => {
    return /* @__PURE__ */ jsx(
      SegmentText,
      {
        segment,
        variableClassName: textStyleCSS({
          fontStyle: "text",
          textStyle: "small"
        })
      },
      segment.id
    );
  }) });
};
const SegmentText = ({
  segment,
  variableClassName
}) => {
  const { getSegment } = useSegmentsCache();
  const intl = useIntl();
  const { data } = useQuery(
    {
      queryKey: ["getSegment", segment.id],
      queryFn: () => getSegment(segment.id),
      notifyOnChangeProps: ["dataUpdatedAt"]
    },
    segmentsClient.queryClient
  );
  let variables = "";
  if (segment.variables && segment.variables.length > 0) {
    variables = `: ${segment.variables[0].values.map((variable) => getReadableValue(variable, intl)).join(", ")}`;
  }
  if (variableClassName) {
    return /* @__PURE__ */ jsxs("span", { children: [
      data?.segment.name ?? segment.id,
      /* @__PURE__ */ jsx("span", { className: variableClassName, children: variables })
    ] });
  }
  return `${data?.segment.name ?? segment.id}${variables}`;
};
const SegmentsPreset = (props) => {
  const shouldReduceMotion = useReducedMotion();
  const {
    uiPreset,
    id,
    isPinned,
    isSelected,
    pinLimitReached
  } = props;
  const { segments } = uiPreset;
  const [isAnimating, setIsAnimating] = useState(false);
  const { setUISegments } = useSegmentsUIContext();
  const { overlayState } = useContext(UISegmentsDispatchContext);
  const { setSegments } = useSegments();
  const pin = useMutation(
    {
      mutationKey: ["pin-segment-preset", id],
      mutationFn: pinSegmentsPreset,
      onMutate: async () => runOptimisticMutation(
        "getSegmentsPresets",
        (prevState) => updateCachePinMutation(prevState, id, segments),
        [
          "pin-segment-preset",
          "unpin-segment-preset",
          id,
          "getSegmentsPresets"
        ]
      ),
      onError: (error, _, context) => handleFailedOptimisticMutation("getSegmentsPresets", error, context),
      onSettled: () => handleSettledOptimisticMutation(
        ["getSegmentsPresets"],
        ["pin-segment-preset", "unpin-segment-preset", id]
      )
    },
    segmentsClient.queryClient
  );
  const unpin = useMutation(
    {
      mutationKey: ["unpin-segment-preset", id],
      mutationFn: unpinSegmentsPreset,
      onMutate: async () => runOptimisticMutation(
        "getSegmentsPresets",
        (prevState) => updateCacheUnpinMutation(prevState, id),
        [
          "pin-segment-preset",
          "unpin-segment-preset",
          id,
          "getSegmentsPresets"
        ]
      ),
      onError: (error, _, context) => handleFailedOptimisticMutation("getSegmentsPresets", error, context),
      onSettled: () => handleSettledOptimisticMutation(
        ["getSegmentsPresets"],
        ["pin-segment-preset", "unpin-segment-preset", id]
      )
    },
    segmentsClient.queryClient
  );
  const applyPreset = useMutation(
    {
      mutationKey: ["apply-segment-preset", id],
      mutationFn: () => {
        const flattenedSegments = segments.map((segment) => {
          if (segment.variables) {
            const flattenedVariables = segment.variables.map((variable) => {
              const values = flattenQueryResultToValues([], variable.values);
              return {
                ...variable,
                values
              };
            });
            return { ...segment, variables: flattenedVariables };
          }
          return segment;
        });
        setUISegments(flattenedSegments, true);
        return setSegments(flattenedSegments);
      },
      onSuccess: async () => {
        const isValid = await areUiSegmentsValid(segments);
        if (isValid) {
          overlayState?.close();
        }
      }
    },
    segmentsClient.queryClient
  );
  const color = isSelected ? "primary" : "neutral";
  const pinningTooltipText = usePinningTooltipText(pinLimitReached, isPinned);
  const icon = isPinned ? /* @__PURE__ */ jsx(PinOffIcon, {}) : /* @__PURE__ */ jsx(PinIcon, {});
  const pinButtonDataDtProperties = `state:${isPinned ? "unpin" : "pin"}`;
  const onClick = () => isPinned ? unpin.mutate(uiPreset.id) : pin.mutate(segments);
  const labelRef = useRef(null);
  const [showTooltip, setShowTooltip] = useState(false);
  const stackingContext = useOverlayStacking(void 0, "overlay");
  return /* @__PURE__ */ jsx(
    motion.div,
    {
      variants: shouldReduceMotion ? {} : expandAndCollapseAnimationVariants,
      initial: "exit",
      animate: "enter",
      exit: "exit",
      className: clsx({
        [animationHelperStyles.isAnimating]: isAnimating
      }),
      onAnimationStart: () => setIsAnimating(true),
      onAnimationComplete: () => setIsAnimating(false),
      children: /* @__PURE__ */ jsxs(
        InputGroup,
        {
          className: clsx([
            styles.inputGroup,
            { [styles.inputGroupSelected]: isSelected }
          ]),
          children: [
            /* @__PURE__ */ jsx(
              Tooltip,
              {
                placement: "left",
                fallbackPlacements: ["top", "bottom", "right"],
                text: /* @__PURE__ */ jsx(PresetTooltipText, { segments }),
                open: showTooltip,
                onOpenChange: (isOpen) => {
                  const hasOverFlow = (labelRef.current?.scrollWidth ?? 0) > (labelRef.current?.clientWidth ?? 0);
                  setShowTooltip(isOpen && hasOverFlow);
                },
                className: textStyleCSS({
                  fontStyle: "text",
                  textStyle: "small-emphasized"
                }),
                style: { zIndex: Number(stackingContext.style.zIndex) + 1 },
                children: /* @__PURE__ */ jsx(
                  Button,
                  {
                    color,
                    width: "100%",
                    textAlign: "start",
                    className: clsx(styles.presetButton, {
                      [styles.selectionMark]: isSelected
                    }),
                    onClick: () => applyPreset.mutate(),
                    children: /* @__PURE__ */ jsx(Button.Label, { className: styles.buttonLabel, ref: labelRef, children: segments.map((segment, index) => /* @__PURE__ */ jsx(
                      SegmentsPresetText,
                      {
                        segment,
                        isFirst: index === 0
                      },
                      segment.id
                    )) })
                  }
                )
              }
            ),
            /* @__PURE__ */ jsx(
              Tooltip,
              {
                placement: "right",
                text: pinningTooltipText,
                className: textStyleCSS({
                  fontStyle: "text",
                  textStyle: "small-emphasized"
                }),
                style: { zIndex: Number(stackingContext.style.zIndex) + 1 },
                children: /* @__PURE__ */ jsx(
                  Button,
                  {
                    color,
                    className: styles.suffixButton,
                    onClick,
                    disabled: pinLimitReached && !isPinned,
                    "aria-label": pinningTooltipText,
                    "data-dt-name": "PinUnpin",
                    "data-dt-properties": pinButtonDataDtProperties,
                    children: icon
                  }
                )
              }
            )
          ]
        }
      )
    }
  );
};
export {
  SegmentsPreset,
  expandAndCollapseAnimationVariants
};
