// @dynatrace/strato-components/charts/xy-chart-adoption/histogram/utils/build-histogram-imperative-handle.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var build_histogram_imperative_handle_exports = {};
__export(build_histogram_imperative_handle_exports, {
  buildHistogramImperativeHandle: () => buildHistogramImperativeHandle
});
module.exports = __toCommonJS(build_histogram_imperative_handle_exports);
var import_serialize_histogram_config = require("./serialize-histogram-config.js");
const buildHistogramImperativeHandle = (xyRef, config) => ({
  get element() {
    return xyRef.current?.element ?? null;
  },
  getConfig: () => (0, import_serialize_histogram_config.serializeHistogramConfig)(config),
  downloadData: () => xyRef.current?.downloadData(),
  exploreMode: () => xyRef.current?.exploreMode(),
  zoomMode: () => xyRef.current?.zoomMode(),
  panMode: () => xyRef.current?.panMode(),
  zoomIn: () => xyRef.current?.zoomIn(),
  zoomOut: () => xyRef.current?.zoomOut(),
  reset: () => xyRef.current?.reset()
});
