// @dynatrace/strato-components/charts/xy-chart-adoption/histogram/utils/derive-xychart-props-from-config.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var derive_xychart_props_from_config_exports = {};
__export(derive_xychart_props_from_config_exports, {
  deriveXYChartPropsFromConfig: () => deriveXYChartPropsFromConfig,
  pickOnZoomChange: () => pickOnZoomChange
});
module.exports = __toCommonJS(derive_xychart_props_from_config_exports);
const pickOnZoomChange = (config, rootOnZoomChange) => rootOnZoomChange ?? config.interactions?.onZoomChange;
const deriveXYChartPropsFromConfig = (config, rootOnZoomChange) => ({
  xAxisConfig: config.xAxis,
  disableZoom: config.interactions?.zoom?.enabled === false,
  disablePan: config.interactions?.pan?.enabled === false,
  toolbarEnabled: config.toolbar?.enabled !== false,
  toolbarPlacement: config.toolbar?.placement,
  toolbarIntents: config.toolbar?.intents,
  onZoomChange: pickOnZoomChange(config, rootOnZoomChange)
});
