// @dynatrace/strato-components/esm/charts/core/components/legend/range/components/RangeLegendSegmentsItem.js
import { jsx } from "react/jsx-runtime";
import { useRef } from "react";
import Colors from "@dynatrace/strato-design-tokens/colors";
import { ORIENTATION } from "../../../../constants.js";
import { DEFAULT_GAP_THICKNESS, DEFAULT_SEGMENT_COLOR } from "../constants.js";
import { useOrientation } from "../context/RangeLegendOrientationContext.js";
import { useScale } from "../context/RangeLegendScaleContext.js";
import { useLegendSegmentKeyboardHandler } from "../hooks/use-legend-segment-keyboard-handler.js";
import { getSegmentItemPositionAndSize } from "../utils/get-segment-item-position-and-size.js";
import { getStrokeDasharray } from "../utils/get-stroke-dasharray.js";
const RangeLegendSegmentsItem = ({
  item,
  segmentSize,
  startPosition
}) => {
  const rectRef = useRef(null);
  const { orientation } = useOrientation();
  const isVertical = orientation === ORIENTATION.Vertical;
  const { scale } = useScale();
  const [, endScalePosition] = scale.range();
  const hasPreviousGap = startPosition > 0;
  const hasNextGap = Math.ceil(startPosition + segmentSize) < endScalePosition;
  const { itemHeight, itemWidth, itemX, itemY } = getSegmentItemPositionAndSize(
    orientation,
    startPosition,
    segmentSize
  );
  const strokeDashArray = getStrokeDasharray(
    { height: itemHeight, width: itemWidth },
    {
      bottom: isVertical && hasNextGap,
      left: !isVertical && hasPreviousGap,
      right: !isVertical && hasNextGap,
      top: isVertical && hasPreviousGap
    }
  );
  const { onBlurHandler, onFocusHandler, onKeyDownHandler } = useLegendSegmentKeyboardHandler(
    rectRef.current,
    item.from,
    item.to,
    startPosition
  );
  return /* @__PURE__ */ jsx(
    "rect",
    {
      ref: rectRef,
      x: itemX,
      y: itemY,
      width: itemWidth,
      height: itemHeight,
      fill: item.color ?? DEFAULT_SEGMENT_COLOR,
      stroke: Colors.Background.Base.Default,
      strokeWidth: DEFAULT_GAP_THICKNESS,
      strokeDasharray: strokeDashArray,
      role: "presentation",
      tabIndex: 0,
      onFocus: onFocusHandler,
      onBlur: onBlurHandler,
      onKeyDown: onKeyDownHandler
    }
  );
};
export {
  RangeLegendSegmentsItem
};
