// @dynatrace/strato-components/esm/charts/core/components/legend/range/components/RangeLegendSegmentsOverlay.js
import { jsx } from "react/jsx-runtime";
import { useCallback, useEffect, useRef } from "react";
import { ORIENTATION } from "../../../../constants.js";
import { OVERLAY_ADDITIONAL_SIZE, OVERLAY_OFFSET } from "../constants.js";
import { useOrientation } from "../context/RangeLegendOrientationContext.js";
import { useLegendSegmentsOverlayHandlers } from "../hooks/use-legend-segments-overlay-handlers.js";
const RangeLegendSegmentsOverlay = (props) => {
  const { itemMarkers } = props;
  const overlayRef = useRef(null);
  const { onMouseLeaveHandler, onMouseMoveHandler } = useLegendSegmentsOverlayHandlers(itemMarkers);
  const { orientation } = useOrientation();
  const isVertical = orientation === ORIENTATION.Vertical;
  const overlaySizeHandler = useCallback(
    (legendBar, overlayEl) => {
      const { height: legendHeight, width: legendWidth } = legendBar.getBoundingClientRect();
      const overlayHeight = legendHeight + (isVertical ? 0 : OVERLAY_ADDITIONAL_SIZE);
      const overlayY = isVertical ? 0 : -OVERLAY_OFFSET;
      const overlayWidth = legendWidth + (isVertical ? OVERLAY_ADDITIONAL_SIZE : 0);
      const overlayX = isVertical ? -OVERLAY_OFFSET : 0;
      overlayEl.setAttribute("y", overlayY.toString());
      overlayEl.setAttribute("x", overlayX.toString());
      overlayEl.setAttribute("height", overlayHeight.toString());
      overlayEl.setAttribute("width", overlayWidth.toString());
    },
    [isVertical]
  );
  useEffect(() => {
    if (overlayRef.current) {
      const legendBar = overlayRef.current.closest("svg")?.previousElementSibling;
      if (legendBar) {
        overlaySizeHandler(legendBar, overlayRef.current);
        const observer = new MutationObserver(() => {
          overlaySizeHandler(legendBar, overlayRef.current);
        });
        observer.observe(legendBar, {
          attributes: true,
          childList: true,
          subtree: true,
          characterData: true
        });
        return () => observer.disconnect();
      }
    }
  }, [overlaySizeHandler]);
  return /* @__PURE__ */ jsx(
    "rect",
    {
      ref: overlayRef,
      role: "presentation",
      id: "segments-overlay",
      fill: "transparent",
      onMouseMove: onMouseMoveHandler,
      onMouseLeave: onMouseLeaveHandler
    }
  );
};
export {
  RangeLegendSegmentsOverlay
};
