// @dynatrace/strato-components/esm/layouts/help-menu/HelpMenu.js
import { jsx, jsxs } from "react/jsx-runtime";
import { useState } from "react";
import { FormattedMessage, useIntl } from "react-intl";
import {
  DescriptionIcon,
  ExternalLinkIcon,
  FeedbackIcon,
  GuideIcon,
  HelpIcon,
  InformationIcon,
  KeyboardIcon,
  PlayIcon,
  WhatsNewIcon
} from "@dynatrace/strato-icons";
import { getAppId, getAppVersion } from "@dynatrace-sdk/app-environment";
import { sendIntent } from "@dynatrace-sdk/navigation";
import { Button } from "../../buttons/button/Button.js";
import { Menu } from "../../navigation/menu/Menu.js";
import { Tooltip } from "../../overlays/tooltip/Tooltip.js";
import { Flex } from "../flex/Flex.js";
const COMPONENT_NAME = "HelpMenu";
const MENU_ITEMS = [
  {
    label: /* @__PURE__ */ jsx(FormattedMessage, { defaultMessage: "What's new", id: "TY3F/kE6n91t0TL0" }),
    icon: /* @__PURE__ */ jsx(WhatsNewIcon, {}),
    key: "whatsNew"
  },
  {
    label: /* @__PURE__ */ jsx(FormattedMessage, { defaultMessage: "Get started", id: "kOQNUENAg02eeaTY" }),
    icon: /* @__PURE__ */ jsx(GuideIcon, {}),
    key: "getStarted"
  },
  {
    label: /* @__PURE__ */ jsx(FormattedMessage, { defaultMessage: "Documentation", id: "npzzIhowJGIZhj8d" }),
    icon: /* @__PURE__ */ jsx(DescriptionIcon, {}),
    key: "documentation"
  },
  {
    label: /* @__PURE__ */ jsx(
      FormattedMessage,
      {
        defaultMessage: "Keyboard shortcuts",
        id: "JmaVR5qUnuxE3qnb"
      }
    ),
    icon: /* @__PURE__ */ jsx(KeyboardIcon, {}),
    key: "keyboardShortcuts"
  },
  {
    label: /* @__PURE__ */ jsx(
      FormattedMessage,
      {
        defaultMessage: "Try in Playground",
        id: "Nar807M4NJhqMw/T"
      }
    ),
    icon: /* @__PURE__ */ jsx(PlayIcon, {}),
    key: "playground"
  },
  {
    label: /* @__PURE__ */ jsx(FormattedMessage, { defaultMessage: "Share feedback", id: "Ia87HCOLBoxMcV9t" }),
    icon: /* @__PURE__ */ jsx(FeedbackIcon, {}),
    key: "feedback"
  },
  {
    label: /* @__PURE__ */ jsx(FormattedMessage, { defaultMessage: "About this app", id: "9/nhbjHpPPAhMbHo" }),
    icon: /* @__PURE__ */ jsx(InformationIcon, {}),
    key: "about"
  }
];
const HelpMenuFooter = () => {
  const version = getAppVersion();
  if (!version) {
    return null;
  }
  return /* @__PURE__ */ jsx(Menu.Label, { children: /* @__PURE__ */ jsx(
    "div",
    {
      style: {
        maxWidth: 500,
        overflow: "hidden",
        textOverflow: "ellipsis",
        whiteSpace: "nowrap"
      },
      children: /* @__PURE__ */ jsx(
        FormattedMessage,
        {
          defaultMessage: "Version {version}",
          description: "App version label",
          id: "W2KEm8AIcM8pOixf",
          values: { version }
        }
      )
    }
  ) });
};
const HelpMenuLink = ({
  href,
  target = "_blank",
  onSelect,
  label,
  icon,
  ...rest
}) => {
  return /* @__PURE__ */ jsxs(
    Menu.Link,
    {
      ...rest,
      onClick: () => onSelect?.(),
      href,
      target,
      children: [
        icon ? /* @__PURE__ */ jsx(Menu.Prefix, { children: icon }) : null,
        /* @__PURE__ */ jsxs(Flex, { justifyContent: "space-between", width: "inherit", children: [
          label,
          /* @__PURE__ */ jsx(ExternalLinkIcon, { style: { flexShrink: 0 } })
        ] })
      ]
    }
  );
};
const HelpMenuButton = ({
  onSelect,
  label,
  icon,
  ...rest
}) => {
  return /* @__PURE__ */ jsxs(Menu.Item, { onSelect, ...rest, children: [
    icon ? /* @__PURE__ */ jsx(Menu.Prefix, { children: icon }) : null,
    label
  ] });
};
const HelpMenu = (props) => {
  const {
    entries,
    onOpenChange,
    children,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow
  } = props;
  const [isOpen, setIsOpen] = useState(false);
  const intl = useIntl();
  const label = intl.formatMessage({
    defaultMessage: "Help",
    id: "xHrhU3DP4e0nge7c"
  });
  const handleOpenChange = (value) => {
    setIsOpen(value);
    onOpenChange?.(value);
  };
  return /* @__PURE__ */ jsxs(
    Menu,
    {
      open: isOpen,
      onOpenChange: handleOpenChange,
      "data-dt-component": COMPONENT_NAME,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      children: [
        /* @__PURE__ */ jsx(Tooltip, { text: label, children: /* @__PURE__ */ jsx(
          Menu.Trigger,
          {
            "data-dt-component": COMPONENT_NAME,
            "data-dt-name": "OpenHelpMenu",
            children: /* @__PURE__ */ jsx(Button, { "aria-label": label, children: /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(HelpIcon, {}) }) })
          }
        ) }),
        /* @__PURE__ */ jsxs(Menu.Content, { children: [
          MENU_ITEMS.map(({ label: label2, key, icon }) => {
            const entry = entries[key];
            if (!entry) {
              return null;
            }
            if (entry === "default") {
              return /* @__PURE__ */ jsx(
                HelpMenuDefaults,
                {
                  entryKey: key,
                  icon,
                  label: label2
                },
                key
              );
            }
            if (Array.isArray(entry)) {
              return /* @__PURE__ */ jsxs(Menu.Sub, { children: [
                /* @__PURE__ */ jsxs(Menu.SubTrigger, { children: [
                  /* @__PURE__ */ jsx(Menu.Prefix, { children: icon }),
                  label2
                ] }),
                /* @__PURE__ */ jsx(Menu.SubContent, { children: entry.map((entryChild) => {
                  if ("href" in entryChild) {
                    return /* @__PURE__ */ jsx(
                      HelpMenuLink,
                      {
                        label: entryChild.label,
                        onSelect: entryChild.onSelect,
                        href: entryChild.href,
                        target: entryChild.target
                      },
                      entryChild.label
                    );
                  }
                  return /* @__PURE__ */ jsx(
                    HelpMenuButton,
                    {
                      label: entryChild.label,
                      onSelect: entryChild.onSelect
                    },
                    entryChild.label
                  );
                }) })
              ] }, key);
            }
            if ("href" in entry) {
              return /* @__PURE__ */ jsx(
                HelpMenuLink,
                {
                  label: label2,
                  icon,
                  onSelect: entry.onSelect,
                  href: entry.href,
                  target: entry.target
                },
                key
              );
            }
            return /* @__PURE__ */ jsx(
              HelpMenuButton,
              {
                label: label2,
                icon,
                onSelect: entry.onSelect
              },
              key
            );
          }),
          children,
          /* @__PURE__ */ jsx(HelpMenuFooter, {})
        ] })
      ]
    }
  );
};
const HelpMenuDefaults = (props) => {
  switch (props.entryKey) {
    case "whatsNew":
      return /* @__PURE__ */ jsx(
        HelpMenuButton,
        {
          onSelect: () => sendIntent(
            {
              appId: getAppId(),
              version: getAppVersion()
            },
            {
              recommendedAppId: "dynatrace.hub",
              recommendedIntentId: "view_app_releases_add_on_intent"
            }
          ),
          label: props.label,
          icon: props.icon
        }
      );
    case "keyboardShortcuts":
      return /* @__PURE__ */ jsx(
        HelpMenuButton,
        {
          onSelect: () => {
            globalThis.dtRuntime?.uiCommands?.showUiCommands?.();
          },
          label: props.label,
          icon: props.icon
        }
      );
    case "playground":
      return /* @__PURE__ */ jsx(
        HelpMenuLink,
        {
          href: `https://wkf10640.apps.dynatrace.com/ui/apps/${getAppId()}`,
          label: props.label,
          icon: props.icon
        }
      );
    case "about":
      return /* @__PURE__ */ jsx(
        HelpMenuButton,
        {
          onSelect: () => {
            sendIntent(
              { appId: getAppId() },
              {
                recommendedAppId: "dynatrace.hub",
                recommendedIntentId: "view_app_details"
              }
            );
          },
          label: props.label,
          icon: props.icon
        }
      );
    default:
      return null;
  }
};
export {
  HelpMenu
};
