// @dynatrace/strato-components/overlays/index.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var overlays_exports = {};
__export(overlays_exports, {
  DismissButton: () => import_useOverlay.DismissButton,
  Modal: () => import_Modal.Modal,
  Overlay: () => import_Overlay.Overlay,
  Sheet: () => import_Sheet.Sheet,
  Tooltip: () => import_Tooltip.Tooltip,
  _Drawer: () => import_Drawer.Drawer,
  _useDialog: () => import_useDialog.useDialog,
  _useOverlay: () => import_useOverlay.useOverlay,
  _useTooltipState: () => import_useTooltipState.useTooltipState,
  basePlacements: () => import_placement.basePlacements,
  useOverlayWithTrigger: () => import_useOverlayWithTrigger.useOverlayWithTrigger
});
module.exports = __toCommonJS(overlays_exports);
var import_Drawer = require("./drawer/Drawer.js");
var import_useDialog = require("./hooks/useDialog.js");
var import_useOverlay = require("./hooks/useOverlay.js");
var import_useOverlayWithTrigger = require("./hooks/useOverlayWithTrigger.js");
var import_useTooltipState = require("./hooks/useTooltipState.js");
var import_Modal = require("./modal/Modal.js");
var import_Overlay = require("./overlay/Overlay.js");
var import_Sheet = require("./sheet/Sheet.js");
var import_Tooltip = require("./tooltip/Tooltip.js");
var import_placement = require("./types/placement.js");
