// @dynatrace/strato-components/esm/charts/honeycomb/components/toolbar/HoneycombToolbar.js
import { jsx } from "react/jsx-runtime";
import { isUndefined } from "lodash-es";
import { HoneycombDownloadDataButton } from "./buttons/HoneycombDownloadDataButton.js";
import { useBehavioralTrackingProps as _useBehavioralTrackingProps } from "../../../../core/hooks/useBehavioralTrackingProps.js";
import { useTooltipReducer } from "../../../core/components/overlay-tooltip/useTooltipReducer.js";
import { ChartToolbarRenderer } from "../../../core/components/toolbar/ChartToolbarRenderer.js";
import { useHoneycombData } from "../../hooks/use-honeycomb-data.js";
import { useHoneycombToolbarConfig } from "../../hooks/use-honeycomb-toolbar-config.js";
import { useHoneycombToolbarVisibleState } from "../../store/selectors.js";
import { buildAndDownloadCsv } from "../../utils/build-and-download-csv.js";
const HoneycombToolbar = () => {
  const toolbarVisible = useHoneycombToolbarVisibleState();
  const dispatch = useTooltipReducer();
  const toolbarConfig = useHoneycombToolbarConfig();
  const rawData = useHoneycombData();
  if (isUndefined(toolbarConfig)) {
    return null;
  }
  const isDownloadEnabled = toolbarConfig?.downloadCSV?.enabled ?? toolbarConfig?.downloadData?.enabled ?? false;
  const defaultToolbarPlacement = "top-right";
  const handleMouseEnter = () => {
    dispatch({ type: "RESET_TOOLTIP" })();
  };
  const handleDownloadDataToolbar = () => buildAndDownloadCsv(rawData);
  const showDefaultToolbarDownloadButton = isDownloadEnabled && !toolbarConfig?.intents;
  const showMenuToolbarButton = isDownloadEnabled && !!toolbarConfig?.intents?.length;
  const initialA11yState = showDefaultToolbarDownloadButton ? [{ download: 0 }] : [{ intent: 0 }];
  const behaviorTrackingProps = _useBehavioralTrackingProps(
    toolbarConfig ?? {}
  )[0];
  return /* @__PURE__ */ jsx(
    ChartToolbarRenderer,
    {
      visible: toolbarVisible,
      collapsed: toolbarConfig.collapsed,
      draggable: true,
      placement: toolbarConfig.placement || defaultToolbarPlacement,
      initialA11yState,
      onMouseEnter: handleMouseEnter,
      menuDisplayMode: "never",
      isDownloadEnabled: showMenuToolbarButton,
      intents: toolbarConfig?.intents,
      handleDownloadDataToolbar,
      removeDragHandle: true,
      ...behaviorTrackingProps,
      children: showDefaultToolbarDownloadButton && /* @__PURE__ */ jsx(HoneycombDownloadDataButton, {})
    }
  );
};
HoneycombToolbar["displayName"] = "HoneycombToolbar";
export {
  HoneycombToolbar
};
