// @dynatrace/strato-components/layouts/page-layout/PageLayout.sty.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var PageLayout_css_exports = {};
__export(PageLayout_css_exports, {
  content: () => content,
  contentContainer: () => contentContainer,
  contentTopLeftRadius: () => contentTopLeftRadius,
  contentTopRightRadius: () => contentTopRightRadius,
  headerContainer: () => headerContainer,
  layoutRoot: () => layoutRoot
});
module.exports = __toCommonJS(PageLayout_css_exports);
var import_AppRoot_css_ts_vanilla = require("../../core/components/app-root/AppRoot.css");
var import_PageLayout_css_ts_vanilla = require("./PageLayout.css");
var content = "_193kgwe3-3-6-0";
var contentContainer = "_193kgwe2-3-6-0";
var contentTopLeftRadius = "_193kgwe4-3-6-0";
var contentTopRightRadius = "_193kgwe5-3-6-0";
var headerContainer = "_193kgwe1-3-6-0";
var layoutRoot = "_193kgwe0-3-6-0";
