// @dynatrace/strato-components/layouts/page-layout/PageLayout.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var PageLayout_exports = {};
__export(PageLayout_exports, {
  PageLayout: () => PageLayout,
  usePageLayoutPortalContext: () => import_PageLayoutPortalContext.usePageLayoutPortalContext
});
module.exports = __toCommonJS(PageLayout_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = require("clsx");
var import_react = require("react");
var import_DetailsLayout = require("./details-layout/DetailsLayout.js");
var styles = __toESM(require("./PageLayout.sty.js"));
var import_PageLayoutPortalContext = require("./shared/PageLayoutPortalContext.js");
var import_portal_slots = require("./shared/portal-slots.js");
var import_slot_defaults = require("./shared/slot-defaults.js");
var import_slot_registry = require("./shared/slot-registry.js");
var import_TeleportTarget = require("./shared/TeleportTarget.js");
var import_useStableHost = require("./shared/useStableHost.js");
var import_SidebarLayout = require("./sidebar-layout/SidebarLayout.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
function resolveOptionalSlot(config, target) {
  if (!config) {
    return null;
  }
  return { ...config, children: target };
}
function MainShell() {
  const { hosts, registry } = (0, import_PageLayoutPortalContext.usePageLayoutPortalContext)();
  const snapshot = (0, import_react.useSyncExternalStore)(
    registry.subscribe,
    registry.getSnapshot
  );
  const sidebarConfig = snapshot.sidebar;
  const detailsConfig = snapshot.details;
  const detailsLayoutContext = snapshot.detailsLayoutContext;
  const showTopLeftRadius = sidebarConfig !== null;
  const showTopRightRadius = detailsConfig !== null && detailsLayoutContext?.mode !== "collapsed";
  const contentClassName = (0, import_clsx.clsx)(
    styles.content,
    showTopLeftRadius && styles.contentTopLeftRadius,
    showTopRightRadius && styles.contentTopRightRadius
  );
  const sidebarTarget = (0, import_react.useMemo)(
    () => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TeleportTarget.TeleportTarget, { host: hosts.sidebar }),
    [hosts.sidebar]
  );
  const detailsTarget = (0, import_react.useMemo)(
    () => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TeleportTarget.TeleportTarget, { host: hosts.details }),
    [hosts.details]
  );
  const contentTarget = (0, import_react.useMemo)(
    () => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TeleportTarget.TeleportTarget, { host: hosts.content }),
    [hosts.content]
  );
  const sidebarSlot = (0, import_react.useMemo)(
    () => resolveOptionalSlot(snapshot.sidebar, sidebarTarget),
    [snapshot.sidebar, sidebarTarget]
  );
  const detailsSlot = (0, import_react.useMemo)(
    () => resolveOptionalSlot(snapshot.details, detailsTarget),
    [snapshot.details, detailsTarget]
  );
  const contentSlot = (0, import_react.useMemo)(() => {
    if (snapshot.content) {
      const { config, domProps, ref } = snapshot.content;
      return {
        children: contentTarget,
        config,
        domProps: {
          ...domProps,
          className: (0, import_clsx.clsx)(contentClassName, domProps.className)
        },
        ref
      };
    }
    return {
      children: void 0,
      config: { minWidth: import_slot_defaults.contentDefaults.minWidth },
      domProps: { className: contentClassName }
    };
  }, [snapshot.content, contentTarget, contentClassName]);
  const sidebarContentSlot = (0, import_react.useMemo)(
    () => ({
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_DetailsLayout.DetailsLayout, { details: detailsSlot, content: contentSlot }),
      config: { minWidth: import_slot_defaults.contentDefaults.minWidth },
      domProps: {}
    }),
    [detailsSlot, contentSlot]
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_SidebarLayout.SidebarLayout, { sidebar: sidebarSlot, content: sidebarContentSlot });
}
function HeaderShell() {
  const { hosts, registry } = (0, import_PageLayoutPortalContext.usePageLayoutPortalContext)();
  const snapshot = (0, import_react.useSyncExternalStore)(
    registry.subscribe,
    registry.getSnapshot
  );
  const headerConfig = snapshot.header;
  if (headerConfig === null) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      ...headerConfig.domProps,
      ref: headerConfig.ref,
      className: (0, import_clsx.clsx)(styles.headerContainer, headerConfig.domProps.className),
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TeleportTarget.TeleportTarget, { host: hosts.header })
    }
  );
}
const LayoutRoot = (0, import_react.memo)(
  (0, import_react.forwardRef)((props, forwardedRef) => {
    const { children, className: rootClassName, ...rootProps } = props;
    const [behavioralTrackingProps, restRootProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(rootProps);
    const headerHost = (0, import_useStableHost.useStableHost)();
    const sidebarHost = (0, import_useStableHost.useStableHost)();
    const contentHost = (0, import_useStableHost.useStableHost)();
    const detailsHost = (0, import_useStableHost.useStableHost)();
    const registry = (0, import_react.useMemo)(() => (0, import_slot_registry.createSlotRegistry)(), []);
    const context = (0, import_react.useMemo)(
      () => ({
        hosts: {
          header: headerHost,
          sidebar: sidebarHost,
          content: contentHost,
          details: detailsHost
        },
        registry
      }),
      // hosts and registry are stable
      // eslint-disable-next-line react-hooks/exhaustive-deps
      []
    );
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_PageLayoutPortalContext.PageLayoutPortalContext.Provider, { value: context, children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      "div",
      {
        ...restRootProps,
        ...behavioralTrackingProps,
        ref: forwardedRef,
        className: (0, import_clsx.clsx)(styles.layoutRoot, rootClassName),
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(HeaderShell, {}),
          /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { className: styles.contentContainer, children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(MainShell, {}),
            children
          ] })
        ]
      }
    ) });
  })
);
LayoutRoot.displayName = "PageLayout";
const PageLayout = Object.assign(LayoutRoot, {
  /** Displays content at the very top of the page. */
  Header: import_portal_slots.PageLayoutHeader,
  /** Displays content in the collapsible, resizable sidebar panel on the left. */
  Sidebar: import_portal_slots.PageLayoutSidebar,
  /**
   * Displays the styled main content area (padding, background, border-radius).
   */
  Content: import_portal_slots.PageLayoutContent,
  /** Displays content in the collapsible, resizable details panel on the right. */
  Details: import_portal_slots.PageLayoutDetailsWithControlBar
});
