// @dynatrace/strato-geo/map/store/map-store.context.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var map_store_context_exports = {};
__export(map_store_context_exports, {
  MapStore: () => MapStore,
  initialState: () => initialState
});
module.exports = __toCommonJS(map_store_context_exports);
var import_react = require("react");
var import_core = require("@dynatrace/strato-components/core");
const initialSelection = { highlightedSeries: void 0, visible: void 0 };
const initialState = {
  tooltip: {
    enabled: true,
    hoveredLayerId: void 0,
    location: void 0
  },
  overlay: {
    mouseInBounds: false
  },
  selection: initialSelection,
  zoom: {
    zoomInEnabled: true,
    zoomOutEnabled: true,
    zoomToFitEnabled: true,
    resetEnabled: false
  }
};
const MapStore = (0, import_react.createContext)(
  (0, import_core.createStore)(initialState)
);
MapStore["displayName"] = "MapStore";
