// @dynatrace/strato-geo/map/store/selectors.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var selectors_exports = {};
__export(selectors_exports, {
  useMapOverlayState: () => useMapOverlayState,
  useMapSelectionState: () => useMapSelectionState,
  useMapTooltipHoveredLayerIdState: () => useMapTooltipHoveredLayerIdState,
  useMapTooltipLocationState: () => useMapTooltipLocationState,
  useMapZoomState: () => useMapZoomState,
  useSelector: () => useSelector
});
module.exports = __toCommonJS(selectors_exports);
var import_lodash_es = require("lodash");
var import_core = require("@dynatrace/strato-components/core");
var import_map_store_context = require("./map-store.context.js");
const useSelector = (selector) => (0, import_core.useStoreSelector)(import_map_store_context.MapStore, selector, import_lodash_es.isEqual);
const useMapSelectionState = () => useSelector(({ selection }) => selection);
const useMapTooltipLocationState = () => useSelector(({ tooltip }) => tooltip.location);
const useMapTooltipHoveredLayerIdState = () => useSelector(({ tooltip }) => tooltip.hoveredLayerId);
const useMapOverlayState = () => useSelector(({ overlay }) => overlay);
const useMapZoomState = () => useSelector(({ zoom }) => zoom);
