// @dynatrace/strato-components/esm/core/hooks/intent.js
import { useReducer } from "react";
import {
  sendIntent,
  sendIntentWithResponse
} from "@dynatrace-sdk/navigation";
function getAppIconURL(appId) {
  if (!appId) {
    return null;
  }
  return `/platform/app-engine/registry/v1/app-icons/${appId}`;
}
function useIntent({ payload, options }) {
  let iconURL = null;
  if (options?.recommendedAppId) {
    iconURL = getAppIconURL(options.recommendedAppId);
  }
  function send() {
    if (!options) {
      return sendIntent(payload);
    }
    return sendIntent(payload, options);
  }
  return { send, iconURL };
}
function reducer(_, action) {
  switch (action.type) {
    case "send":
      return { status: "pending" };
    case "success":
      return { status: "success", data: action.payload };
    case "error":
      return { status: "error", error: action.payload };
    default:
      throw new Error();
  }
}
function useIntentWithResponse({
  payload,
  options
}) {
  const iconURL = getAppIconURL(options.recommendedAppId);
  const [state, dispatch] = useReducer(reducer, {
    status: "idle"
  });
  async function send(overridePayload) {
    dispatch({ type: "send" });
    const res = await sendIntentWithResponse(
      overridePayload || payload,
      options
    );
    if (!res) {
      const error = new Error("No response received");
      dispatch({
        type: "error",
        payload: error
      });
      return {
        status: "error",
        error
      };
    }
    if (res.ok) {
      dispatch({ type: "success", payload: { response: res?.data } });
      return { status: "success", data: { response: res?.data } };
    } else {
      const error = new Error(`Intent Error: ${res.error.reason}`);
      dispatch({
        type: "error",
        payload: error
      });
      return {
        status: "error",
        error
      };
    }
  }
  return { send, state, iconURL };
}
export {
  useIntent,
  useIntentWithResponse
};
