// @dynatrace/strato-components/esm/core/hooks/usePermissions.js
import { useQuery } from "@tanstack/react-query";
import { effectivePermissionsClient } from "@dynatrace-sdk/client-platform-management-service";
import { permissionsQueryClient } from "../utils/permissionsQueryClient.js";
async function resolvePermissions(permissions) {
  const result = await effectivePermissionsClient.resolveEffectivePermissions({
    body: {
      permissions: permissions.map((permission) => ({ permission }))
    }
  });
  return (result ?? []).map(({ permission, granted }) => ({
    permission,
    granted
  }));
}
function usePermissions(permissions) {
  const queryKey = [
    "permissions",
    ...permissions.slice().sort((a, b) => a.localeCompare(b))
  ];
  const { data, isLoading, isError } = useQuery(
    {
      queryKey,
      queryFn: () => resolvePermissions(permissions),
      enabled: permissions.length > 0
    },
    permissionsQueryClient
  );
  const details = data ?? [];
  const isAllowed = !isLoading && !isError && (permissions.length === 0 || details.length > 0 && details.every(
    ({ granted }) => granted === "true" || granted === "condition"
  ));
  return { isAllowed, isLoading, details };
}
export {
  usePermissions
};
