// @dynatrace/strato-components/esm/typography/list/List.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import {
  Children,
  forwardRef,
  isValidElement
} from "react";
import { listCSS, listItemCSS } from "./List.sty.js";
import { useAriaLabelingProps } from "../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useSafeForwardProps } from "../../core/hooks/useSafeForwardProps.js";
import { Text } from "../text/Text.js";
const COMPONENT_NAME = "List";
function interleaveListItems(listItems) {
  const items = [];
  Children.forEach(listItems, (listItem) => {
    if (isValidElement(listItem)) {
      if (listItem.type === List) {
        if (items.length === 0) {
          items.push([listItem]);
        } else {
          items[items.length - 1].push(listItem);
        }
      } else {
        items.push([listItem]);
      }
    }
  });
  return items.map((values, index) => {
    const itemKey = `list-item-${getItemKey(values, index)}`;
    return /* @__PURE__ */ jsx("li", { className: listItemCSS, children: values }, itemKey);
  });
}
function getItemKey(values, index) {
  const isArray = Array.isArray(values);
  const value = isArray ? values?.[0] : values;
  if (isValidElement(value)) {
    if (value.key) {
      return value.key;
    } else if (value.props && value.props.children) {
      return getItemKey(value.props.children, index);
    }
    return value;
  }
  if (value === void 0) {
    return index;
  }
  return value;
}
const List = /* @__PURE__ */ forwardRef(
  (props, ref) => {
    const {
      children,
      ordered = void 0,
      textStyle,
      fontStyle,
      start = 1,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      className: consumerClassName,
      style: consumerStyle,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
    const ariaLabelingProps = useAriaLabelingProps(restProps);
    const forwardProps = useSafeForwardProps(restProps, ariaLabelingProps);
    return /* @__PURE__ */ jsx(
      Text,
      {
        ref,
        as: ordered ? "ol" : "ul",
        ...ordered && {
          start
        },
        textStyle,
        fontStyle,
        className: clsx(listCSS, consumerClassName),
        style: consumerStyle,
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        ...ariaLabelingProps,
        ...behavioralTrackingProps,
        ...forwardProps,
        "data-dt-component": COMPONENT_NAME,
        children: interleaveListItems(children)
      }
    );
  }
);
List.displayName = COMPONENT_NAME;
export {
  List
};
