// @dynatrace/strato-components/tables/DataTable/components/DataTableHeaderCell.sty.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DataTableHeaderCell_css_exports = {};
__export(DataTableHeaderCell_css_exports, {
  childHeader: () => childHeader,
  columnSizingHandle: () => columnSizingHandle,
  customHeaderDensity: () => customHeaderDensity,
  firstColumn: () => firstColumn,
  firstHeaderRow: () => firstHeaderRow,
  firstRightPinnedColumn: () => firstRightPinnedColumn,
  firstUnpinnedColumn: () => firstUnpinnedColumn,
  headerCell: () => headerCell,
  headerCellActiveInteraction: () => headerCellActiveInteraction,
  headerCellEllipsis: () => headerCellEllipsis,
  headerWidthVar: () => headerWidthVar,
  hovered: () => hovered,
  isMultiSortedHeader: () => isMultiSortedHeader,
  isSortedHeader: () => isSortedHeader,
  lastHeaderRow: () => lastHeaderRow,
  lastLeftPinnedColumn: () => lastLeftPinnedColumn,
  leftAdditionalElevationVar: () => leftAdditionalElevationVar,
  leftPositionVar: () => leftPositionVar,
  narrow: () => narrow,
  parentHeader: () => parentHeader,
  rightPositionVar: () => rightPositionVar,
  sortableHeader: () => sortableHeader,
  subRowIndicatorColumn: () => subRowIndicatorColumn
});
module.exports = __toCommonJS(DataTableHeaderCell_css_exports);
var import_AppRoot_css_ts_vanilla = require("../../../core/components/app-root/AppRoot.css");
var import_DataTable_css_ts_vanilla = require("../DataTable.css");
var import_Alignment_css_ts_vanilla = require("./Alignment.css");
var import_ColumnOrder_css_ts_vanilla = require("../features/ColumnOrder/ColumnOrder.css");
var import_Thresholds_css_ts_vanilla = require("../features/Thresholds/Thresholds.css");
var import_DataTableHeaderCell_css_ts_vanilla = require("./DataTableHeaderCell.css");
var import_createRuntimeFn = require("@vanilla-extract/recipes/createRuntimeFn");
var childHeader = "_11pvd3z3-3-6-0";
var columnSizingHandle = (0, import_createRuntimeFn.createRuntimeFn)({ defaultClassName: "_11pvd3zl-3-6-0", variantClassNames: { active: { true: "_11pvd3zm-3-6-0", false: "_11pvd3zn-3-6-0" }, last: { true: "_11pvd3zo-3-6-0", false: "_11pvd3zp-3-6-0" }, pinnedColumn: { false: "_11pvd3zq-3-6-0", left: "_11pvd3zr-3-6-0", right: "_11pvd3zs-3-6-0" } }, defaultVariants: {}, compoundVariants: [] });
var customHeaderDensity = { "default": "_11pvd3za-3-6-0", condensed: "_11pvd3zb-3-6-0", comfortable: "_11pvd3zc-3-6-0" };
var firstColumn = "_11pvd3zd-3-6-0";
var firstHeaderRow = "_11pvd3z0-3-6-0";
var firstRightPinnedColumn = "_11pvd3zg-3-6-0";
var firstUnpinnedColumn = "_11pvd3zf-3-6-0";
var headerCell = (0, import_createRuntimeFn.createRuntimeFn)({ defaultClassName: "_11pvd3zt-3-6-0", variantClassNames: { pinnedColumn: { false: "_11pvd3zu-3-6-0", left: "_11pvd3zv-3-6-0", right: "_11pvd3zw-3-6-0" }, rowDensity: { customHeaderRenderer: "_11pvd3zx-3-6-0", "default": "_11pvd3zy-3-6-0", condensed: "_11pvd3zz-3-6-0", comfortable: "_11pvd3z10-3-6-0" }, verticalDividers: { true: "_11pvd3z11-3-6-0", false: "_11pvd3z12-3-6-0" }, isDragAndDropDisabled: { true: "_11pvd3z13-3-6-0", false: "_11pvd3z14-3-6-0" } }, defaultVariants: { rowDensity: "default", verticalDividers: false }, compoundVariants: [] });
var headerCellActiveInteraction = "_11pvd3z16-3-6-0";
var headerCellEllipsis = "_11pvd3z15-3-6-0";
var headerWidthVar = "var(--_11pvd3zj-3-6-0)";
var hovered = "_11pvd3z4-3-6-0";
var isMultiSortedHeader = "_11pvd3z8-3-6-0";
var isSortedHeader = "_11pvd3z7-3-6-0";
var lastHeaderRow = "_11pvd3z1-3-6-0";
var lastLeftPinnedColumn = "_11pvd3ze-3-6-0";
var leftAdditionalElevationVar = "var(--_11pvd3zk-3-6-0)";
var leftPositionVar = "var(--_11pvd3zh-3-6-0)";
var narrow = "_11pvd3z5-3-6-0";
var parentHeader = "_11pvd3z2-3-6-0";
var rightPositionVar = "var(--_11pvd3zi-3-6-0)";
var sortableHeader = "_11pvd3z6-3-6-0";
var subRowIndicatorColumn = "_11pvd3z9-3-6-0";
