// @dynatrace/strato-components/tables/DataTable/components/DataTableBody.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DataTableBody_exports = {};
__export(DataTableBody_exports, {
  DataTableBody: () => DataTableBody
});
module.exports = __toCommonJS(DataTableBody_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_DataTableRow = require("./DataTableRow.js");
var import_DataTableContext = require("../contexts/DataTableContext.js");
var import_RowDetailsElement = require("../features/RowDetails/RowDetailsElement.js");
var import_RowOrder = require("../features/RowOrder/RowOrder.js");
var styles = __toESM(require("./DataTableBody.sty.js"));
var import_constants = require("../constants.js");
function DataTableBody(props) {
  const { rows, rowVirtualizer, columnVirtualizer } = props;
  const table = (0, import_DataTableContext.useDataTable)();
  const queuedMeasurementRowIdRef = (0, import_react.useRef)(/* @__PURE__ */ new Set());
  const rowMeasure = (0, import_react.useCallback)(
    (node) => {
      if (!node || !(node instanceof HTMLElement)) {
        return;
      }
      const rowId = node.getAttribute("data-rowid");
      if (rowId === null || queuedMeasurementRowIdRef.current.has(rowId)) {
        return;
      }
      queuedMeasurementRowIdRef.current.add(rowId);
      requestAnimationFrame(() => {
        if (node.isConnected) {
          rowVirtualizer.measureElement(node);
        }
        queuedMeasurementRowIdRef.current.delete(rowId);
      });
    },
    [rowVirtualizer]
  );
  const columnMeasure = (0, import_react.useCallback)(
    (node) => {
      if (!node || !(node instanceof HTMLElement)) {
        return;
      }
      columnVirtualizer.measureElement(node);
    },
    [columnVirtualizer]
  );
  const firstRenderedRowIndex = rows.at(0)?.index;
  const rowSeparation = table.getVisualVariantOptions().rowSeparation;
  const virtualColumns = columnVirtualizer.getVirtualItems();
  const draggingRowId = table.getState().draggingRowInfo?.id;
  const draggingRowIdExpanded = `${draggingRowId}${import_constants.DATA_TABLE_ROW_EXPANDED_ID_SUFFIX}`;
  const rowElements = rowVirtualizer.getVirtualItems().map((virtualRow, index) => {
    const originalRow = rows[virtualRow.index];
    let parity = virtualRow.index % 2 === 0 ? "even" : "odd";
    const isDraggedRow = draggingRowId === void 0 ? false : originalRow.id === draggingRowId || originalRow.id === draggingRowIdExpanded;
    const detailsParity = originalRow[import_constants.DATA_TABLE_DETAILS_ROW_PARITY];
    if (detailsParity !== void 0) {
      parity = detailsParity;
    }
    if (originalRow[import_constants.DATA_TABLE_DETAILS_ROW_IDENTIFIER] === true) {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_RowDetailsElement.RowDetailsElement,
        {
          originalRow,
          virtualRow,
          virtualItemIndex: index,
          rowSeparation,
          rowParity: parity,
          isLastRowOnPage: virtualRow.index === rows.length - 1,
          rowMeasure,
          isDraggedRow,
          children: table.options.rowDetailsTemplate?.({
            row: originalRow.original
          })
        },
        originalRow.id
      );
    }
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_DataTableRow.DataTableRow,
      {
        row: originalRow,
        virtualRow,
        virtualItemIndex: index,
        rowCount: rows.length,
        virtualColumns,
        isDraggedRow,
        isInDragOverlay: false,
        rowParity: parity,
        rowMeasure,
        columnMeasure: (
          // For columns, only measure the first rendered row.
          // All other rows will have the same column widths.
          firstRenderedRowIndex === virtualRow.index ? columnMeasure : void 0
        ),
        scrollToIndex: rowVirtualizer.scrollToIndex
      },
      `row-${virtualRow.key}`
    );
  });
  const bodyContent = table.options.enableRowOrderDragAndDrop ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_RowOrder.RowOrderingDragAndDropContext,
    {
      rows,
      virtualColumns,
      rowSeparation,
      children: rowElements
    }
  ) : (
    // If row order drag and drop is not enabled then don't use the RowOrderingDragAndDropContext
    // eslint-disable-next-line react/jsx-no-useless-fragment
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, { children: rowElements })
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      role: "rowgroup",
      "data-rowgroup": "body",
      className: (0, import_clsx.default)(styles.body, {
        [styles.loading]: table.getState().loading
      }),
      children: bodyContent
    }
  );
}
DataTableBody.displayName = "DataTableBody";
