// @dynatrace/strato-components/tables/DataTable/components/DataTableHeaderCell.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DataTableHeaderCell_exports = {};
__export(DataTableHeaderCell_exports, {
  DataTableHeaderCell: () => DataTableHeaderCell,
  getHeaderAriaLabel: () => getHeaderAriaLabel
});
module.exports = __toCommonJS(DataTableHeaderCell_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_table = require("@tanstack/react-table");
var import_dynamic = require("@vanilla-extract/dynamic");
var import_clsx = __toESM(require("clsx"));
var import_react2 = require("react");
var import_react_dom = require("react-dom");
var import_react_intl = require("react-intl");
var alignmentStyles = __toESM(require("./Alignment.sty.js"));
var import_DataTablePinningShadowHeaderCells = require("./ColumnPinning/DataTablePinningShadowHeaderCells.js");
var headerCellStyles = __toESM(require("./DataTableHeaderCell.sty.js"));
var import_useMergeRefs = require("../../../core/hooks/useMergeRefs.js");
var import_Tooltip = require("../../../overlays/tooltip/Tooltip.js");
var import_TextEllipsis = require("../../../typography/text-ellipsis/TextEllipsis.js");
var import_data_table_border_width = require("../../utils/data-table-border-width.js");
var import_get_column_label = require("../../utils/get-column-label.js");
var import_constants = require("../constants.js");
var import_DataTableContext = require("../contexts/DataTableContext.js");
var tableStyles = __toESM(require("../DataTable.sty.js"));
var columnOrderStyles = __toESM(require("../features/ColumnOrder/ColumnOrder.sty.js"));
var import_SortingArrowIndicators = require("../features/Sorting/SortingArrowIndicators.js");
var import_SubRowExpanders = require("../features/SubRows/SubRowExpanders.js");
var import_SubRows = require("../features/SubRows/SubRows.js");
var import_InlineUserActionsMenu = require("../features/UserActions/InlineUserActionsMenu.js");
var import_useHeaderCellInteractions = require("../hooks/useHeaderCellInteractions.js");
var import_useHeaderDragAndDrop = require("../hooks/useHeaderDragAndDrop.js");
var import_messages = require("../messages.js");
const HeaderTextEllipsis = (0, import_react2.memo)(
  ({
    children,
    onTextOverflow
  }) => {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_TextEllipsis.TextEllipsis,
      {
        onTextOverflow,
        className: headerCellStyles.headerCellEllipsis,
        children
      }
    );
  }
);
function getHeaderAriaLabel({
  intl,
  columnLabel,
  nextSortingOrder,
  canSort,
  columnDragAndDropMode,
  isCustomHeader,
  role
}) {
  if (role === "presentation") {
    return void 0;
  }
  if (canSort && columnDragAndDropMode === "none") {
    return intl.formatMessage(import_messages.dataTableMessages.sortingLabel, {
      column: columnLabel,
      direction: nextSortingOrder
    });
  }
  if (canSort && columnDragAndDropMode === "enabled") {
    return intl.formatMessage(import_messages.dataTableMessages.sortingDragAndDropLabel, {
      column: columnLabel,
      direction: nextSortingOrder
    });
  }
  if (columnDragAndDropMode !== "none") {
    return intl.formatMessage(import_messages.dataTableMessages.dragAndDropLabel, {
      column: columnLabel
    });
  }
  return isCustomHeader ? columnLabel : void 0;
}
function getHeaderAriaSortDescription({
  intl,
  role,
  isSorted,
  multiSortPriority,
  sortedColumnsCount
}) {
  if (role === "presentation" || !isSorted) {
    return void 0;
  }
  if (multiSortPriority !== void 0) {
    return intl.formatMessage(import_messages.dataTableMessages.sortingStateWithPriorityLabel, {
      direction: isSorted,
      priority: multiSortPriority,
      total: sortedColumnsCount
    });
  }
  return intl.formatMessage(import_messages.dataTableMessages.sortingStateLabel, {
    direction: isSorted
  });
}
function getInteractionDtName({
  columnDragAndDropMode,
  canSort
}) {
  if (columnDragAndDropMode === "only" || columnDragAndDropMode === "enabled" && !canSort) {
    return "ColumnHeaderOrder";
  }
  if (canSort && columnDragAndDropMode === "enabled") {
    return "ColumnHeaderOrderAndSort";
  }
  return "ColumnHeaderSort";
}
function getInteractionDtProperties(canSort, nextSortingOrder) {
  if (!canSort || nextSortingOrder === void 0) {
    return void 0;
  }
  switch (nextSortingOrder) {
    case "asc":
      return "state:ascending";
    case "desc":
      return "state:descending";
    case false:
      return "state:clearSort";
  }
}
function getLeftAdditionalElevation(isPinnedLeft, canResize, columnId, leftPinnedColumns) {
  if (!isPinnedLeft || !canResize) {
    return "0";
  }
  const idx = leftPinnedColumns.findIndex((c) => c.id === columnId);
  return `${leftPinnedColumns.length - 1 - idx}`;
}
function DataTableHeaderCell(props) {
  const {
    header,
    headers,
    headerGroupId,
    rowIndex,
    columnIndex: virtualColumnIndex,
    lastRowIndex
  } = props;
  const intl = (0, import_react_intl.useIntl)();
  const table = (0, import_DataTableContext.useDataTable)();
  const { rowDensity, verticalDividers } = table.getVisualVariantOptions();
  const columnDragAndDropMode = header.getColumnDragAndDropMode();
  const gridPosition = header.getGridPosition(headers);
  const column = header.column;
  const { id: columnId, parent, columns: childColumns, columnDef } = column;
  const columnIndex = rowIndex === 0 ? gridPosition.gridColumnStart - 1 : virtualColumnIndex;
  const canSort = column.getCanSort();
  const canColumnActions = column.getCanColumnActions();
  const textOverflowRef = (0, import_react2.useRef)(false);
  const handleTextOverflow = (0, import_react2.useCallback)((ellipsized) => {
    textOverflowRef.current = ellipsized;
  }, []);
  const { setNodeRef, isDragging, portalRef, dragAndDropProps } = (0, import_useHeaderDragAndDrop.useHeaderCellDragAndDrop)({
    headerId: header.id,
    column,
    columnIndex,
    columnDragAndDropMode
  });
  const divRef = (0, import_react2.useRef)(null);
  const mergedRef = (0, import_useMergeRefs.useMergeRefs)([setNodeRef, divRef]);
  const tableState = table.getState();
  const isActiveFloatingMenu = tableState.floatingMenuInfo?.rowId === null && tableState.floatingMenuInfo?.headerId === header.id;
  const {
    handleCellMouseEnter,
    handleCellMouseLeave,
    handleCellActionsClose,
    handleCellClick
  } = (0, import_useHeaderCellInteractions.useHeaderCellInteractions)({
    headerCell: divRef.current,
    canSort,
    canColumnActions,
    isActiveFloatingMenu,
    rowDensity,
    getToggleSortingHandler: column.getToggleSortingHandler
  });
  if (header.isPlaceholder) {
    return null;
  }
  const headerGroupsLength = table.getHeaderGroups().length;
  const activeSortEntries = tableState.sorting?.filter((sort) => {
    const col = table.getColumn(sort.id);
    return col?.getCanSort() ?? false;
  }) ?? [];
  const sortedColumnsCount = activeSortEntries.length;
  const enableExpanding = table.options.enableExpanding;
  const canResize = column.getCanResize();
  const isResizing = column.getIsResizing();
  const isSorted = column.getIsSorted();
  const columnSortIndex = activeSortEntries.findIndex(
    (sort) => sort.id === column.id
  );
  const multiSortPriority = sortedColumnsCount > 1 && columnSortIndex >= 0 ? columnSortIndex + 1 : void 0;
  const isColumnDragActive = tableState.draggedHeaderColumn !== void 0;
  const isRightAligned = columnDef.alignment === "right";
  const renderVerticalDividers = verticalDividers || table.getHeaderGroups().length > 1 || table.options.enableColumnResizing && (canResize || // when resizing is enabled we need to render the vertical dividers for the spacer column as well
  column.id === import_constants.DATA_TABLE_SPACER_COLUMN_ID);
  const hasCustomHeaderRenderer = columnDef.meta?.hasCustomHeaderRenderer;
  const children = hasCustomHeaderRenderer ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_DataTableContext.DataTableColumnContextProvider, { value: header.column, children: (0, import_react_table.flexRender)(columnDef.header, header.getContext()) }) : (0, import_react_table.flexRender)(columnDef.header, header.getContext());
  const isStringHeader = typeof children === "string";
  const role = columnDef.meta?.spacer === true ? "presentation" : "columnheader";
  const columnLabel = (0, import_get_column_label.getColumnLabel)(column);
  const nextSortingOrder = column.getNextSortingOrder();
  const ariaLabel = getHeaderAriaLabel({
    intl,
    columnLabel,
    nextSortingOrder,
    canSort,
    columnDragAndDropMode,
    isCustomHeader: !isStringHeader,
    role
  });
  const ariaSortDescription = getHeaderAriaSortDescription({
    intl,
    role,
    isSorted,
    multiSortPriority,
    sortedColumnsCount
  });
  const isSubRowIndicatorColumn = enableExpanding === true && table.getSubRowIndicatorColumnId() === column.id;
  const subRowsWidthCorrection = (
    // only render the subrow width correction in the subrow indicator column
    enableExpanding === true && isSubRowIndicatorColumn ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_SubRows.SubRowsHeaderWidthCorrection,
      {
        subRowOpenDepth: table.getSubRowOpenDepth()
      }
    ) : null
  );
  const pinningPosition = header.getIsPinned();
  const isPinnedLeft = pinningPosition === "left";
  const isPinnedRight = pinningPosition === "right";
  const leftPinnedPosition = isPinnedLeft ? header.getLeftPinnedPosition() : void 0;
  const leftPinnedPositionPx = leftPinnedPosition === void 0 ? void 0 : `${leftPinnedPosition}px`;
  const leftAdditionalElevation = getLeftAdditionalElevation(
    isPinnedLeft,
    canResize,
    columnId,
    table.getLeftVisibleLeafColumns()
  );
  const rightPinnedPosition = isPinnedRight ? header.getRightPinnedPosition() : void 0;
  const rightPinnedPositionPx = rightPinnedPosition === void 0 ? void 0 : `${rightPinnedPosition}px`;
  const hasChildren = childColumns.length > 0;
  const isEdgeColumnInGroup = parent?.getCenterVisibleLeafColumns().at(-1) === column || parent?.getLeftVisibleLeafColumns().at(-1) === column || parent?.getRightVisibleLeafColumns().at(0) === column;
  const isLastVisibleLeftPinnedHeader = isPinnedLeft ? header.getIsLastVisibleLeftPinnedHeader() : false;
  const isFirstVisibleRightPinnedHeader = isPinnedRight ? header.getIsFirstVisibleRightPinnedHeader() : false;
  const isFirstVisibleUnpinnedHeader = pinningPosition === false ? header.getIsFirstVisibleUnpinnedHeader() : false;
  const isLastUnpinnedColumnWithoutSpacer = columnId === table.getCenterVisibleLeafColumns().at(-2)?.id;
  const edgeAdjustedGridPosition = {
    gridColumnStart: gridPosition.gridColumnStart,
    gridColumnEnd: gridPosition.gridColumnEnd,
    gridRowStart: isEdgeColumnInGroup ? 1 : gridPosition.gridRowStart,
    gridRowEnd: isEdgeColumnInGroup ? `span ${headerGroupsLength}` : gridPosition.gridRowEnd
  };
  const interactionDtName = getInteractionDtName({
    columnDragAndDropMode,
    canSort
  });
  const interactionDtProperties = getInteractionDtProperties(
    canSort,
    nextSortingOrder
  );
  const renderSubRowHeaderExpander = () => {
    if (!enableExpanding || !isSubRowIndicatorColumn || hasCustomHeaderRenderer) {
      return null;
    }
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_SubRowExpanders.SubRowsHeaderExpander, { isAllRowsExpanded: table.getIsAllRowsExpanded() });
  };
  let headerContent = /* @__PURE__ */ (0, import_react.createElement)(
    "div",
    {
      ...dragAndDropProps,
      key: `Header-${columnId}`,
      ref: mergedRef,
      role,
      "data-headerid": header?.id,
      "data-header-row": rowIndex,
      "data-column": columnIndex,
      "data-columnid": columnId,
      "data-columntype": columnDef.columnType,
      "data-dt-name": interactionDtName,
      "data-dt-properties": interactionDtProperties,
      className: (0, import_clsx.default)(
        headerCellStyles.headerCell({
          rowDensity: hasCustomHeaderRenderer ? "customHeaderRenderer" : rowDensity,
          pinnedColumn: pinningPosition,
          verticalDividers: renderVerticalDividers,
          isDragAndDropDisabled: header.getIsDragAndDropDisabled()
        }),
        alignmentStyles.horizontalAlignment[columnDef.alignment ?? import_constants.DATA_TABLE_DEFAULT_HORIZONTAL_ALIGNMENT],
        alignmentStyles.verticalAlignment[table.getHeaderVerticalAlignment()],
        {
          [headerCellStyles.firstColumn]: columnIndex === 0,
          [headerCellStyles.firstHeaderRow]: rowIndex === 0,
          [headerCellStyles.lastLeftPinnedColumn]: isLastVisibleLeftPinnedHeader,
          [headerCellStyles.firstUnpinnedColumn]: isFirstVisibleUnpinnedHeader,
          [headerCellStyles.firstRightPinnedColumn]: isFirstVisibleRightPinnedHeader,
          [headerCellStyles.lastHeaderRow]: rowIndex === lastRowIndex,
          [headerCellStyles.parentHeader]: hasChildren,
          [headerCellStyles.childHeader]: !!parent,
          [headerCellStyles.sortableHeader]: canSort,
          [headerCellStyles.isSortedHeader]: isSorted,
          [headerCellStyles.isMultiSortedHeader]: multiSortPriority !== void 0,
          [tableStyles.prefixColumn]: columnDef.meta?.builtinPlacement === "prefix",
          [headerCellStyles.headerCellActiveInteraction]: isActiveFloatingMenu,
          [headerCellStyles.customHeaderDensity[rowDensity]]: hasCustomHeaderRenderer,
          [headerCellStyles.subRowIndicatorColumn]: isSubRowIndicatorColumn,
          // add grab cursor styling only if column drag and drop is enabled but sorting is not
          [columnOrderStyles.headerDragAndDrop.enabled]: !isDragging && columnDragAndDropMode !== "none" && !header.getCanSortByHeaderClick(),
          [columnOrderStyles.headerDragAndDrop.inactive]: tableState.draggedHeaderColumn !== void 0
        }
      ),
      "aria-sort": header.getAriaSortAttribute(),
      "aria-description": ariaSortDescription,
      "data-sort-desc-first": columnDef.sortDescFirst ?? false,
      "aria-label": ariaLabel,
      style: {
        ...(0, import_dynamic.assignInlineVars)({
          [headerCellStyles.leftPositionVar]: leftPinnedPositionPx,
          [headerCellStyles.rightPositionVar]: rightPinnedPositionPx,
          [headerCellStyles.leftAdditionalElevationVar]: leftAdditionalElevation
        }),
        ...gridPosition
      },
      onMouseEnter: handleCellMouseEnter,
      onMouseLeave: handleCellMouseLeave,
      onClick: handleCellClick
    },
    renderSubRowHeaderExpander(),
    isRightAligned && subRowsWidthCorrection,
    isStringHeader ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(HeaderTextEllipsis, { onTextOverflow: handleTextOverflow, children }) : children,
    canSort && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_SortingArrowIndicators.SortingArrowIndicators,
      {
        isSorted,
        nextSortingOrder,
        multiSortPriority,
        rowDensity
      }
    ),
    canColumnActions && !isColumnDragActive && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_DataTableContext.DataTableHeaderContextProvider, { value: header, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_InlineUserActionsMenu.InlineUserActionsMenu,
      {
        headerGroupOrRowId: headerGroupId,
        columnId,
        type: "header",
        "aria-label": intl.formatMessage(import_messages.dataTableMessages.columnActionLabel),
        "data-header-row": rowIndex,
        "data-column": columnIndex,
        onClose: handleCellActionsClose
      }
    ) }),
    !isRightAligned && subRowsWidthCorrection,
    canResize && !isColumnDragActive && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "button",
      {
        type: "button",
        "aria-label": intl.formatMessage(import_messages.dataTableMessages.resizeColumnLabel, {
          column: columnLabel
        }),
        "data-header-row": rowIndex,
        "data-column": columnIndex,
        "data-trigger": "resize",
        className: (0, import_clsx.default)(
          headerCellStyles.columnSizingHandle({
            active: isResizing,
            last: isLastUnpinnedColumnWithoutSpacer,
            pinnedColumn: pinningPosition
          }),
          {
            [headerCellStyles.firstRightPinnedColumn]: isFirstVisibleRightPinnedHeader
          }
        ),
        ...isEdgeColumnInGroup && {
          style: {
            // if this is the last column of a header group, the resize handle should stretch the full header height
            height: table.getTotalHeaderHeight() - (0, import_data_table_border_width.getDataTableBorderWidth)()
          }
        },
        onMouseDown: header.handleResize,
        onTouchStart: header.handleResize,
        onKeyUp: header.handleKeyboardResize,
        onDoubleClick: header.handleResizeExpand,
        "data-dt-name": "ColumnHeaderResize"
      },
      `Header-${columnId}-resize-handler`
    )
  );
  if (isDragging) {
    headerContent = (0, import_react_dom.createPortal)(
      headerContent,
      portalRef.current || document.body
    );
  }
  if (isStringHeader) {
    headerContent = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Tooltip.Tooltip, { disabled: !textOverflowRef.current, text: children, children: headerContent });
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    headerContent,
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_DataTablePinningShadowHeaderCells.DataTablePinningShadowHeaderCells,
      {
        isLeftShadow: isLastVisibleLeftPinnedHeader && !hasChildren,
        isRightShadow: isFirstVisibleRightPinnedHeader && !hasChildren,
        leftPosition: leftPinnedPosition,
        rightPosition: rightPinnedPosition,
        gridPosition: edgeAdjustedGridPosition
      }
    )
  ] });
}
DataTableHeaderCell.displayName = "DataTableHeaderCell";
