// @dynatrace/strato-components/tables/DataTable/components/DataTableDefaultCell.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DataTableDefaultCell_exports = {};
__export(DataTableDefaultCell_exports, {
  DataTableDefaultCell: () => DataTableDefaultCell
});
module.exports = __toCommonJS(DataTableDefaultCell_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_constants = require("../constants.js");
var alignmentStyles = __toESM(require("./Alignment.sty.js"));
var cellStyles = __toESM(require("./DataTableCell.sty.js"));
var defaultCellStyles = __toESM(require("./DataTableDefaultCell.sty.js"));
var import_DataTableContext = require("../contexts/DataTableContext.js");
var import_SubRowExpanders = require("../features/SubRows/SubRowExpanders.js");
const DataTableDefaultCell = (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    children,
    style: consumerStyle,
    className: consumerClassName,
    "data-testid": dataTestId,
    ...remainingProps
  } = props;
  const table = (0, import_DataTableContext.useDataTable)();
  const column = (0, import_DataTableContext.useDataTableColumn)();
  const isHeader = (0, import_DataTableContext.useDataTableIsHeaderContextProvider)();
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    "div",
    {
      ref: forwardedRef,
      className: (0, import_clsx.default)(
        defaultCellStyles.defaultCell({
          rowDensity: table.getVisualVariantOptions().rowDensity
        }),
        alignmentStyles.horizontalAlignment[column.columnDef.alignment ?? import_constants.DATA_TABLE_DEFAULT_HORIZONTAL_ALIGNMENT],
        consumerClassName
      ),
      style: consumerStyle,
      "data-testid": dataTestId,
      ...remainingProps,
      children: [
        table.options.enableExpanding === true && table.getSubRowIndicatorColumnId() === column.columnDef.id && (isHeader ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_SubRowExpanders.SubRowsHeaderExpander,
          {
            isAllRowsExpanded: table.getIsAllRowsExpanded()
          }
        ) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_SubRowExpanders.SubRowsCellExpander, { columnType: column.columnDef.columnType })),
        !column.columnDef.meta?.builtin ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          "div",
          {
            className: (0, import_clsx.default)(
              cellStyles.innerCell({
                lineWrap: !isHeader && column.getIsLineWrapped()
              })
            ),
            children
          }
        ) : children
      ]
    }
  );
});
DataTableDefaultCell.displayName = "DataTable.DefaultCell";
