// @dynatrace/strato-components/tables/DataTable/components/DataTableContent.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DataTableContent_exports = {};
__export(DataTableContent_exports, {
  DataTableContent: () => DataTableContent
});
module.exports = __toCommonJS(DataTableContent_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_intl = require("react-intl");
var import_DataTableBody = require("./DataTableBody.js");
var import_DataTableHeader = require("./DataTableHeader.js");
var import_DataTableContext = require("../contexts/DataTableContext.js");
var import_FloatingMenu = require("../features/UserActions/FloatingMenu.js");
var import_messages = require("../messages.js");
var import_DataTableScrollShadow = require("./DataTableScrollShadow.js");
var import_data_table_border_width = require("../../utils/data-table-border-width.js");
const DataTableContent = (props) => {
  const intl = (0, import_react_intl.useIntl)();
  const table = (0, import_DataTableContext.useDataTable)();
  const visibleLeafColumnsWithoutBuiltin = table.getVisibleLeafColumnsWithoutBuiltin();
  const hasVisibleColumns = visibleLeafColumnsWithoutBuiltin.length > 0;
  if (!hasVisibleColumns) {
    return null;
  }
  const { rowVirtualizer, columnVirtualizer } = props;
  const rows = table.getRowModel().rows;
  const draggingRow = table.getState().draggingRowInfo;
  const isDraggingRow = !!draggingRow;
  const isDataEmpty = rowVirtualizer.options.count === 0;
  const isColumnsEmpty = columnVirtualizer.options.count === 0;
  const isEmpty = isDataEmpty || isColumnsEmpty;
  const scrollOffset = columnVirtualizer.scrollOffset ?? 0;
  const scrollElement = columnVirtualizer.scrollElement;
  const tableViewportWidth = scrollElement?.clientWidth ?? 0;
  const scrollElementWidth = scrollElement?.scrollWidth ?? 0;
  const visualVariantOptions = table.getVisualVariantOptions();
  const hasOverflowLeft = scrollOffset > 0;
  const gridHeight = isEmpty ? "auto" : rowVirtualizer.getTotalSize();
  const borderWidth = visualVariantOptions.contained ? (0, import_data_table_border_width.getDataTableBorderWidth)() : 0;
  const hasOverflowRight = scrollElementWidth - scrollOffset - borderWidth > tableViewportWidth;
  const cellActionAriaLabel = intl.formatMessage(
    import_messages.dataTableMessages.cellActionLabel
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    "div",
    {
      role: "grid",
      style: {
        display: "grid",
        height: gridHeight,
        gridTemplateColumns: table.getGridTemplateColumns(),
        gridTemplateRows: table.getGridTemplateRows()
      },
      onMouseOver: (event) => {
        if (!isDraggingRow) {
          table.handleInteractiveRowMouseOver(event);
        }
      },
      onMouseLeave: (event) => {
        table.handleInteractiveRowMouseLeave(event);
      },
      onMouseDown: (event) => {
        if (
          // main button (left button)
          event.button === 0 || // auxiliary button (wheel/middle button)
          event.button === 1
        ) {
          table.handleInteractiveRowClick(event);
        }
      },
      onContextMenu: (event) => {
        if (event.target instanceof Element && event.target.closest("a")) {
          return;
        }
        table.getUserActionFloatingMenuContextMenuHandler()(event);
      },
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
          import_DataTableContext.DataTableContentContextProvider,
          {
            value: {
              hasOverflowLeft,
              hasOverflowRight,
              cellActionAriaLabel
            },
            children: [
              visualVariantOptions.headers === "default" && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_DataTableHeader.DataTableHeader, {}),
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_DataTableBody.DataTableBody,
                {
                  rows,
                  rowVirtualizer,
                  columnVirtualizer
                }
              ),
              (table.options.enableCellActions || table.options.enableColumnActions || table.options.enableSorting) && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_FloatingMenu.CellActionFloatingMenu, {})
            ]
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_DataTableScrollShadow.DataTableScrollShadow,
          {
            hasOverflowLeft,
            hasOverflowRight
          }
        )
      ]
    }
  );
};
