// @dynatrace/strato-components/tables/DataTable/components/DataTableHeader.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DataTableHeader_exports = {};
__export(DataTableHeader_exports, {
  DataTableHeader: () => DataTableHeader
});
module.exports = __toCommonJS(DataTableHeader_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var headerStyles = __toESM(require("./DataTableHeader.sty.js"));
var import_DataTableHeaderCell = require("./DataTableHeaderCell.js");
var import_DataTableContext = require("../contexts/DataTableContext.js");
var import_ColumnOrderDragAndDropContext = require("../features/ColumnOrder/ColumnOrderDragAndDropContext.js");
function DataTableHeader() {
  const table = (0, import_DataTableContext.useDataTable)();
  const headerGroups = table?.getHeaderGroups();
  const headerRef = (0, import_react.useRef)(null);
  (0, import_react.useEffect)(() => {
    if (!headerRef.current) {
      return;
    }
    const resizeObserver = new ResizeObserver((entries) => {
      const height = entries.at(0)?.borderBoxSize?.at(0)?.blockSize;
      table.setObservedHeaderHeight(height ?? 0);
    });
    resizeObserver.observe(headerRef.current);
    return () => {
      resizeObserver.disconnect();
    };
  }, [table]);
  const headerRows = headerGroups.map((headerGroup, rowIndex, rows) => {
    const headers = table.getOrderedHeaderGroups()[rowIndex];
    const headerCells = headers.map((header, columnIndex) => {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_DataTableHeaderCell.DataTableHeaderCell,
        {
          header,
          headers,
          headerGroupId: headerGroup.id,
          rowIndex,
          columnIndex,
          lastRowIndex: rows.length - 1
        },
        header.id
      );
    });
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { role: "row", className: headerStyles.headerRow, children: headerCells }, headerGroup.id);
  });
  const headerContent = table.options.enableColumnOrdering ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ColumnOrderDragAndDropContext.ColumnOrderDragAndDropContext, { children: headerRows }) : (
    // If column ordering is not enabled, don't use the ColumnOrderDragAndDropContext
    // eslint-disable-next-line react/jsx-no-useless-fragment
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, { children: headerRows })
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_DataTableContext.DataTableIsHeaderContextProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      ref: headerRef,
      role: "rowgroup",
      "data-rowgroup": "header",
      className: headerStyles.header,
      children: headerContent
    }
  ) });
}
DataTableHeader.displayName = "DataTableHeader";
