// @dynatrace/strato-components/tables/DataTable/components/DataTableRow.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DataTableRow_exports = {};
__export(DataTableRow_exports, {
  DataTableRow: () => DataTableRow
});
module.exports = __toCommonJS(DataTableRow_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_sortable = require("@dnd-kit/sortable");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_is_target_interactive = require("../../utils/is-target-interactive.js");
var import_DataTableContext = require("../contexts/DataTableContext.js");
var tableStyles = __toESM(require("../DataTable.sty.js"));
var import_DataTablePinningShadowCells = require("./ColumnPinning/DataTablePinningShadowCells.js");
var import_DataTableCell = require("./DataTableCell.js");
var styles = __toESM(require("./DataTableRow.sty.js"));
var import_RowDetails_css = require("../features/RowDetails/RowDetails.sty.js");
var import_RowInteractivity = require("../features/RowInteractivity/RowInteractivity.js");
var import_SubRowCellContext = require("../features/SubRows/SubRowCellContext.js");
const ROW_LONG_PRESS_TIMEOUT_DURATION = 500;
function DataTableRow(props) {
  const {
    row,
    virtualRow,
    rowCount,
    virtualColumns,
    isDraggedRow,
    isInDragOverlay,
    virtualItemIndex,
    rowParity,
    rowMeasure,
    columnMeasure,
    scrollToIndex: scrollToIndexProp
  } = props;
  const table = (0, import_DataTableContext.useDataTable)();
  const virtualRowIndex = virtualRow.index;
  const hasSubRows = table.options.enableExpanding === true;
  const subRowIndicatorColumnId = hasSubRows ? table.getSubRowIndicatorColumnId() : null;
  const subRowDepth = hasSubRows ? row.getSubRowDepth() : 0;
  const thresholdPill = row.getThresholdPill();
  const tableState = table.getState();
  const floatingMenuInfoRowId = tableState.floatingMenuInfo?.rowId;
  const floatingMenuInfoColumnId = tableState.floatingMenuInfo?.columnId;
  const scrollToIndex = scrollToIndexProp ?? (() => void 0);
  const { rowInteractivityClassName, rowInteractivityProps } = (0, import_RowInteractivity.useRowInteractivityProps)(row, scrollToIndex);
  const rowLink = row.getLinkForRow();
  const rowLinkRef = (0, import_react.useRef)(null);
  const { setNodeRef } = (0, import_sortable.useSortable)({
    id: row.id,
    data: {
      row,
      rowIndex: virtualRowIndex,
      virtualRow
    }
  });
  const isLongPress = (0, import_react.useRef)(false);
  const longPressTimeout = (0, import_react.useRef)(null);
  const handleOpenRowLinkOnEnter = (event) => {
    if (!rowLink) {
      return;
    }
    const isRowEnter = event.target.getAttribute("role") === "row" && event.key === "Enter";
    if (isRowEnter) {
      rowLinkRef.current?.click();
    }
  };
  const handleClickOnRowLink = (event) => {
    const { target } = event;
    if (isLongPress.current || !(target instanceof HTMLElement) && !(target instanceof SVGElement)) {
      return;
    }
    const row2 = target.closest('[role="row"]');
    if (row2 === null || (0, import_is_target_interactive.isTargetInteractive)(row2, target)) {
      return;
    }
    const nativeEvent = event.nativeEvent;
    const newEvent = new MouseEvent("click", {
      ...event.nativeEvent,
      // For some reason the Ctrl and Meta key as well as the button key have to be manually set,
      // even though the native event already contains the correct values
      ctrlKey: nativeEvent.ctrlKey === true,
      metaKey: nativeEvent.metaKey === true,
      button: nativeEvent.button
    });
    rowLinkRef.current?.dispatchEvent(newEvent);
  };
  const leftPinnedColumns = table.getLeftVisibleLeafColumns();
  const lastLeftPinnedColumn = leftPinnedColumns.at(-1);
  const lastLeftPinnedVirtualColumnIndex = leftPinnedColumns.length - 1;
  const rightPinnedColumns = table.getRightVisibleLeafColumns();
  const firstRightPinnedColumn = rightPinnedColumns.at(0);
  const firstRightPinnedVirtualColumnIndex = virtualColumns[virtualColumns.length - rightPinnedColumns.length]?.index;
  const gridRowPosition = row.getGridRowPosition(virtualItemIndex);
  const visibleCells = row.getVisibleCells();
  let linkHandlingEvents = {};
  if (rowLink) {
    linkHandlingEvents = {
      onPointerUp: handleClickOnRowLink,
      onKeyUp: handleOpenRowLinkOnEnter
    };
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      "div",
      {
        role: "row",
        ref: rowMeasure,
        "data-rowid": row.id,
        "data-row": virtualRowIndex,
        className: (0, import_clsx.default)(
          styles.row({
            rowSeparation: table.getVisualVariantOptions().rowSeparation,
            rowParity,
            isDraggedRow
          }),
          {
            [tableStyles.firstBodyRow]: virtualRowIndex === 0,
            [tableStyles.lastBodyRow]: virtualRowIndex === rowCount - 1,
            [`${import_RowDetails_css.rowDetailsExpandedRowCell}`]: row.getIsDetailsOpen(),
            [rowInteractivityClassName]: !isInDragOverlay
          }
        ),
        style: {
          ...gridRowPosition
        },
        onTouchStart: () => {
          longPressTimeout.current = setTimeout(() => {
            isLongPress.current = true;
          }, ROW_LONG_PRESS_TIMEOUT_DURATION);
        },
        onTouchEnd: () => {
          if (longPressTimeout.current !== null) {
            clearTimeout(longPressTimeout.current);
          }
          isLongPress.current = false;
        },
        ...!isInDragOverlay && {
          ...rowInteractivityProps,
          ...linkHandlingEvents
        },
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_DataTableContext.DataTableRowContextProvider, { value: row, children: virtualColumns.map((virtualCell) => {
            const virtualCellIndex = virtualCell.index;
            const cell = visibleCells[virtualCellIndex];
            const column = cell.column;
            const hasActiveFloatingMenu = floatingMenuInfoRowId === row.id && floatingMenuInfoColumnId === column.id;
            return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_SubRowCellContext.SubRowCellContextProvider,
              {
                isSubRowIndicatorColumn: hasSubRows && subRowIndicatorColumnId === column.id,
                cell,
                children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  import_DataTableCell.DataTableCell,
                  {
                    cell,
                    virtualColumnIndex: virtualCellIndex,
                    virtualRowIndex,
                    isInDragOverlay,
                    hasActiveFloatingMenu,
                    columnMeasure,
                    dragAndDropRef: setNodeRef,
                    subRowDepth,
                    thresholdPill,
                    rowData: row.original,
                    cellDeps: column.columnDef?.meta?.cellDeps?.(cell).toString()
                  }
                )
              },
              `${virtualRow.key}-${virtualCell.key}`
            );
          }) }),
          rowLink && // this link element solely holds the link for interactive rows and is not visible to the user
          // it's not directly clickable; instead, the link is opened by interactions with the row (click, enter)
          // eslint-disable-next-line jsx-a11y/anchor-has-content
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "a",
            {
              ref: rowLinkRef,
              href: rowLink,
              onPointerDown: (event) => event.stopPropagation(),
              style: { display: "none" }
            }
          )
        ]
      }
    ),
    !isDraggedRow && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_DataTablePinningShadowCells.DataTablePinningShadowCells,
      {
        lastLeftPinnedColumn: {
          column: lastLeftPinnedColumn,
          index: lastLeftPinnedVirtualColumnIndex
        },
        firstRightPinnedColumn: {
          column: firstRightPinnedColumn,
          index: firstRightPinnedVirtualColumnIndex
        },
        gridRowPosition,
        rowId: row.id
      }
    )
  ] });
}
