// @dynatrace/strato-components/editors/dql-editor/DQLEditor.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DQLEditor_exports = {};
__export(DQLEditor_exports, {
  DQLEditor: () => DQLEditor
});
module.exports = __toCommonJS(DQLEditor_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_intl = require("react-intl");
var actionsMenuStyles = __toESM(require("./actions-menu/ActionsMenu.sty.js"));
var import_ActionsMenu = require("./actions-menu/ActionsMenu.js");
var import_dql_language = require("./dql/dql-language.js");
var import_useValidationCallbacks = require("./useValidationCallbacks.js");
var import_useValidityInfo = require("./useValidityInfo.js");
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useSafeForwardProps = require("../../core/hooks/useSafeForwardProps.js");
var import_is_slot = require("../../core/utils/is-slot.js");
var import_BaseCodeEditor = require("../base-code-editor/BaseCodeEditor.js");
const COMPONENT_NAME = "DQLEditor";
const DQLEditorFwd = /* @__PURE__ */ (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const intl = (0, import_react_intl.useIntl)();
    const defaultPlaceholder = intl.formatMessage({
      defaultMessage: "Add DQL",
      id: "Ir0YCuBX6vOM4t0J",
      description: "Default placeholder for the DQLEditor."
    });
    const {
      id,
      value = "",
      controlState,
      onChange,
      folding,
      onFoldingChange,
      defaultFolding,
      fullHeight = false,
      lineWrap = false,
      placeholder = defaultPlaceholder,
      readOnly = false,
      spellCheck = false,
      onValidationStart,
      onValidationEnd,
      onValidityChange,
      onFocus,
      onBlur,
      size = "default",
      required,
      children,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
    const ariaLabelingProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(restProps);
    const forwardProps = (0, import_useSafeForwardProps.useSafeForwardProps)(restProps, ariaLabelingProps);
    const internalValueRef = (0, import_react.useRef)(value);
    (0, import_react.useEffect)(() => {
      internalValueRef.current = value;
    }, [value]);
    const handleChange = (0, import_react.useCallback)(
      (newValue) => {
        internalValueRef.current = newValue;
        onChange?.(newValue);
      },
      [onChange]
    );
    const getInternalValue = (0, import_react.useCallback)(() => internalValueRef.current, []);
    const [editorConfig, setEditorConfig] = (0, import_react.useState)();
    const [fireValidationStart, fireValidationEnd] = (0, import_useValidationCallbacks.useValidationCallbacks)(
      onValidationStart,
      onValidationEnd
    );
    const setValidityInfo = (0, import_useValidityInfo.useValidityInfo)(onValidityChange);
    (0, import_react.useEffect)(() => {
      setEditorConfig(
        (0, import_dql_language.getDqlConfig)(
          readOnly,
          fireValidationStart,
          fireValidationEnd,
          setValidityInfo
        )
      );
    }, [readOnly, setValidityInfo, fireValidationStart, fireValidationEnd]);
    let hasActionsMenu = false;
    import_react.Children.forEach(children, (child) => {
      if ((0, import_is_slot.isSlot)(child, import_ActionsMenu.ActionsMenu)) {
        hasActionsMenu = true;
      }
    });
    const editorClassName = hasActionsMenu ? (0, import_clsx.default)(
      consumerClassName,
      // Apply right padding on .cm-editor to prevent overlap with the menu trigger
      actionsMenuStyles.editorWithActionsMenu({ size })
    ) : consumerClassName;
    const editor = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_BaseCodeEditor.BaseCodeEditor,
      {
        id,
        ref: forwardedRef,
        onChange: handleChange,
        placeholder,
        controlState,
        readOnly,
        defaultFolding,
        lineWrap,
        value,
        spellCheck,
        fullHeight,
        folding,
        onFoldingChange,
        configuration: editorConfig,
        onFocus,
        onBlur,
        size,
        required,
        className: editorClassName,
        style: consumerStyle,
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        ...ariaLabelingProps,
        ...behavioralTrackingProps,
        ...forwardProps,
        "data-dt-component": COMPONENT_NAME
      }
    );
    if (hasActionsMenu) {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { className: actionsMenuStyles.editorWrapper({ fullHeight }), children: [
        editor,
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ActionsMenu.ActionsMenuInternal, { getValue: getInternalValue, size })
      ] });
    }
    return editor;
  }
);
const DQLEditor = Object.assign(DQLEditorFwd, {
  ActionsMenu: import_ActionsMenu.ActionsMenu
});
