// @dynatrace/strato-components/esm/charts/xy-chart/components/overlay/hooks/useOverlayEvents.js
import { useApplyPanDelta } from "./useApplyPanDelta.js";
import { useApplyZoomDelta } from "./useApplyZoomDelta.js";
import { useResetInteractions } from "./useResetInteractions.js";
import { useZoomToSelection } from "./useZoomToSelection.js";
import { useGridPlotSize } from "../../../../core/components/chart-grid/hooks/useGridPlotSize.js";
import {
  INTERACTION_MODE,
  isSelectCancelled
} from "../../../../core/components/interactions-overlay/types.js";
import { useTooltipActions } from "../../../../core/components/overlay-tooltip/useTooltipActions.js";
import { useSynchronizedCrosshair } from "../../../../core/components/shared-chart-interactions/hooks/useSynchronizedCrosshair.js";
import { useSynchronizedSelection } from "../../../../core/components/shared-chart-interactions/hooks/useSynchronizedSelection.js";
import { ZOOM_CHANGE_TYPE } from "../../../../core/types/zoom.js";
import { getPercentageFromPixel } from "../../../../core/utils/get-percentage-from-pixel.js";
import { useAreaData } from "../../../hooks/useAreaData.js";
import { useBarData } from "../../../hooks/useBarData.js";
import { useDotData } from "../../../hooks/useDotData.js";
import { useLineData } from "../../../hooks/useLineData.js";
import { useOverlayPerformance } from "../../../hooks/useOverlayPerformance.js";
import { useRectData } from "../../../hooks/useRectData.js";
import { useSyncedXScaleByPriority } from "../../../hooks/useSyncedXScaleByPriority.js";
import { useXYChartData } from "../../../hooks/useXYChartData.js";
import { useXYChartId } from "../../../hooks/useXYChartId.js";
import { reduceInteractionMode } from "../../../store/reduce-interaction-mode.js";
import {
  reduceExploreCancelAll,
  reduceExploreCancelHighlight,
  reduceExploreCancelSelect,
  reduceExplorePin,
  reduceExploreReset,
  reduceExploreSelect,
  reduceExploreUnPin
} from "../../../store/reducer-explorer.js";
import {
  useInteractionModeState,
  useSetState,
  useZoomState
} from "../../../store/selectors.js";
import { getPrevailingShapeType } from "../utils/get-prevailing-shape-type.js";
const getHoveredDatapoint = (position, dotSeriesData, lineSeriesData, areaSeriesData, rectData, barData, queryClosest, queryHovered) => {
  const results = {
    dot: queryClosest(dotSeriesData, position.relative),
    line: queryClosest(lineSeriesData, position.relative),
    area: queryClosest(areaSeriesData, position.relative),
    rect: queryHovered(rectData, position.relative),
    bar: queryHovered(barData, position.relative)
  };
  const prevailingShape = getPrevailingShapeType(
    results.dot?.distance,
    results.line?.distance,
    results.area?.distance,
    !!results.rect,
    !!results.bar
  );
  if (!prevailingShape) {
    return void 0;
  }
  return results[prevailingShape];
};
const useOverlayEvents = () => {
  const { filteredSeriesWithDatapoints: rectSeriesData } = useRectData();
  const { filteredSeriesWithDatapoints: barSeriesData } = useBarData();
  const { filteredSeriesWithDatapoints: dotSeriesData } = useDotData();
  const { filteredSeriesWithDatapoints: lineSeriesData } = useLineData();
  const { filteredSeriesWithDatapoints: areaSeriesData } = useAreaData();
  const zoomToSelection = useZoomToSelection();
  const setState = useSetState();
  const mode = useInteractionModeState();
  const applyPanDelta = useApplyPanDelta();
  const applyReset = useResetInteractions();
  const applyZoomDelta = useApplyZoomDelta();
  const zoomState = useZoomState();
  const { queryClosest, queryHovered } = useOverlayPerformance();
  const { axesMetadata } = useXYChartData();
  const chartId = useXYChartId();
  const synchronizedCrosshair = useSynchronizedCrosshair();
  const { width } = useGridPlotSize();
  const {
    updateSelectionState: updateSynchronizedSelection,
    resetSelectionState: resetSynchronizedSelection
  } = useSynchronizedSelection();
  const syncedXScale = useSyncedXScaleByPriority();
  const { hover, pin, unpin, reset, outOfBounds } = useTooltipActions();
  const {
    updateHoverState: updateSynchronizedCrosshair,
    setDisabled: disableSynchronizedCrosshair
  } = synchronizedCrosshair ?? {};
  const resetTooltip = () => {
    setState(reduceExploreCancelHighlight);
    reset();
  };
  const handleModeChange = (event) => {
    const canPan = axesMetadata.isPanEnabled && zoomState.zoomOutEnabled;
    const shouldSwitchToPanMode = event.mode === "pan" && canPan;
    const shouldSwitchToZoomMode = event.mode === "zoom" && axesMetadata.isZoomEnabled;
    if (shouldSwitchToPanMode || shouldSwitchToZoomMode) {
      resetSynchronizedSelection?.();
      setState(reduceInteractionMode(event.mode));
      setState(reduceExploreReset);
    } else if (event.mode === "explore") {
      setState(reduceInteractionMode(INTERACTION_MODE.Explore));
    }
  };
  const handleHover = ({
    position
  }) => {
    if (syncedXScale) {
      const hoveredDate = syncedXScale.invert(position.relative.x);
      disableSynchronizedCrosshair?.(false);
      updateSynchronizedCrosshair?.(chartId, hoveredDate.valueOf());
    }
    const hoveredDatapointQueryResult = getHoveredDatapoint(
      position,
      dotSeriesData,
      lineSeriesData,
      areaSeriesData,
      rectSeriesData,
      barSeriesData,
      queryClosest,
      queryHovered
    );
    const payload = hoveredDatapointQueryResult ? {
      geometry: hoveredDatapointQueryResult.type,
      header1: void 0,
      series: [],
      metadata: hoveredDatapointQueryResult
    } : void 0;
    hover(
      { clientX: position.absolute.x, clientY: position.absolute.y },
      payload
    );
  };
  const handleTogglePin = (shouldPin, position) => {
    if (shouldPin) {
      handlePin(position);
      updateSynchronizedCrosshair?.(void 0);
    } else {
      handleUnPin(position);
    }
  };
  const handleUnPin = (position) => {
    setState(reduceExploreUnPin);
    if (position) {
      const hoveredDatapointQueryResult = getHoveredDatapoint(
        position,
        dotSeriesData,
        lineSeriesData,
        areaSeriesData,
        rectSeriesData,
        barSeriesData,
        queryClosest,
        queryHovered
      );
      unpin();
      if (hoveredDatapointQueryResult) {
        hover(
          { clientX: position.absolute.x, clientY: position.absolute.y },
          {
            geometry: hoveredDatapointQueryResult.type,
            header1: void 0,
            series: [],
            metadata: hoveredDatapointQueryResult
          }
        );
      }
    } else {
      resetTooltip();
    }
  };
  const handlePin = (position) => {
    if (!position) {
      return;
    }
    const hoveredDatapointQueryResult = getHoveredDatapoint(
      position,
      dotSeriesData,
      lineSeriesData,
      areaSeriesData,
      rectSeriesData,
      barSeriesData,
      queryClosest,
      queryHovered
    );
    if (hoveredDatapointQueryResult) {
      setState(reduceExplorePin(hoveredDatapointQueryResult.series.id));
      pin(
        { clientX: position.absolute.x, clientY: position.absolute.y },
        {
          geometry: hoveredDatapointQueryResult.type,
          header1: void 0,
          series: [],
          metadata: hoveredDatapointQueryResult
        }
      );
    }
  };
  const handleReset = (event) => {
    const { mousePosition: position } = event;
    applyReset({ position });
    disableSynchronizedCrosshair?.(false);
    resetSynchronizedSelection?.();
    if (!position) {
      setState(reduceExploreCancelAll);
      return;
    }
    setState(reduceExploreCancelAll);
    const hoveredDatapointQueryResult = getHoveredDatapoint(
      position,
      dotSeriesData,
      lineSeriesData,
      areaSeriesData,
      rectSeriesData,
      barSeriesData,
      queryClosest,
      queryHovered
    );
    if (hoveredDatapointQueryResult) {
      hover(
        { clientX: position.absolute.x, clientY: position.absolute.y },
        {
          geometry: hoveredDatapointQueryResult.type,
          header1: void 0,
          series: [],
          metadata: hoveredDatapointQueryResult
        }
      );
    } else {
      outOfBounds();
    }
  };
  const handleLeave = (_) => {
    updateSynchronizedCrosshair?.(void 0);
    outOfBounds();
  };
  const handlePanChange = (event) => {
    applyPanDelta(event.delta);
  };
  const handleSelectStart = (event) => {
    if (mode === "explore") {
      disableSynchronizedCrosshair?.(true);
      updateSynchronizedSelection?.();
    }
    resetTooltip();
  };
  const handleSelectChange = (event) => {
    const variant = event.mode === "explore" ? "select" : "zoom";
    setState(
      reduceExploreSelect({
        start: event.start,
        end: event.end,
        finished: false,
        accessible: false,
        variant,
        selectionPercentage: getPercentageFromPixel(
          [event.start, event.end],
          width
        )
      })
    );
    if (mode === "explore" && syncedXScale) {
      updateSynchronizedSelection?.({
        selection: [event.start, event.end],
        chartId,
        scale: syncedXScale
      });
    }
  };
  const handleSelectEnd = (event, ref) => {
    if (isSelectCancelled(event)) {
      setState(reduceExploreCancelSelect);
      resetSynchronizedSelection?.();
    } else if (mode === "explore") {
      setState(
        reduceExploreSelect({
          start: event.start,
          end: event.end,
          finished: true,
          accessible: false,
          variant: "select",
          selectionPercentage: getPercentageFromPixel(
            [event.start, event.end],
            width
          )
        })
      );
      ref.current?.focus();
      if (syncedXScale) {
        updateSynchronizedSelection?.({
          selection: [event.start, event.end],
          chartId,
          scale: syncedXScale
        });
      }
    } else {
      zoomToSelection();
    }
  };
  const handleZoomChange = (event) => {
    const canZoomIn = event.behavior === ZOOM_CHANGE_TYPE.ZoomIn && zoomState?.zoomInEnabled;
    const canZoomOut = event.behavior === ZOOM_CHANGE_TYPE.ZoomOut && zoomState.zoomOutEnabled;
    const canZoomX = event.behavior === ZOOM_CHANGE_TYPE.ZoomX;
    const canZoom = canZoomIn || canZoomOut || canZoomX;
    if (canZoom && axesMetadata.isZoomEnabled) {
      setState(reduceExploreCancelAll);
      applyZoomDelta(event.deltaStart, event.deltaEnd, event.behavior);
    }
  };
  return {
    handleModeChange,
    handleHover,
    handleTogglePin,
    handleReset,
    handleLeave,
    handlePanChange,
    handleSelectChange,
    handleSelectEnd,
    handleZoomChange,
    handleSelectStart,
    resetTooltip
  };
};
export {
  getHoveredDatapoint,
  useOverlayEvents
};
