// @dynatrace/strato-geo/esm/map/components/MapContent.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import {
  Layer,
  Map,
  Source
} from "@vis.gl/react-maplibre";
import { isUndefined } from "lodash-es";
import {
  forwardRef,
  useContext,
  useRef,
  useState
} from "react";
import { _useGraphSize as useGraphSize } from "@dynatrace/strato-components/charts";
import { useMergeRefs } from "@dynatrace/strato-components/core";
import {
  BASE_LAYER_FILL_ID,
  BASE_LAYER_LINE_ID,
  COUNTRY_BORDER_WIDTH,
  DEFAULT_BOUNDARIES_BORDER_COLOR,
  DEFAULT_COUNTRIES_FILL_COLOR,
  MAX_LATITUDE,
  MAX_LONGITUDE,
  MIN_LATITUDE,
  MIN_LONGITUDE,
  REGION_BORDER_WIDTH
} from "../constants.js";
import { useMapZoomState, useSetStateZoom } from "../store/store.js";
import { MapToolbar } from "./toolbar/MapToolbar.js";
import { LayerIdsContext } from "../contexts/layer-ids.context.js";
import { MapDataBoundingBoxContext } from "../contexts/map-data-bounding-box.context.js";
import { useActiveInteraction } from "../hooks/use-active-interaction.js";
import { useHoverInteraction } from "../hooks/use-hover-interaction.js";
import { useMapBaseLayerFeatures } from "../hooks/use-map-base-layer-features.js";
import { useMapConfig } from "../hooks/use-map-config.js";
import { useMapLoading } from "../hooks/use-map-loading.js";
import { useMapPerformance } from "../hooks/use-map-performance.js";
import { useOverlayEvents } from "../hooks/use-overlay-events.js";
import { useTooltipEventListeners } from "../hooks/use-tooltip-event-listeners.js";
import { useWebGLContextError } from "../hooks/use-webgl-context-error.js";
import { defaultStyle } from "../map-styles/default-style.js";
import { MapViewImperativeHandler } from "../providers/imperative-handler.provider.js";
import { getColorFromToken } from "../utils/get-color-from-token.js";
import { getDataLayersBoundingBox } from "../utils/get-data-layers-bounding-box.js";
import { toMapBoxInitialViewState } from "../utils/to-mapbox-initial-view-state.js";
import { useSetMapInitialConfiguration } from "../utils/use-set-map-initial-configuration.js";
import { generateNewZoomState } from "./toolbar/utils/zoom-state-utils.js";
const lineLayer = {
  id: BASE_LAYER_LINE_ID,
  type: "line"
};
const fillLayer = {
  id: BASE_LAYER_FILL_ID,
  type: "fill"
};
const MapSource = () => {
  const baseLayerFeatures = useMapBaseLayerFeatures();
  return baseLayerFeatures && /* @__PURE__ */ jsxs(Source, { id: "data", type: "geojson", data: baseLayerFeatures, children: [
    /* @__PURE__ */ jsx(
      Layer,
      {
        ...fillLayer,
        paint: {
          "fill-color": getColorFromToken(DEFAULT_COUNTRIES_FILL_COLOR)
        }
      }
    ),
    /* @__PURE__ */ jsx(
      Layer,
      {
        ...lineLayer,
        type: "line",
        filter: [
          "any",
          ["==", "region_type", "COUNTRY"],
          ["==", "region_type", "REGION"]
        ],
        paint: {
          "line-color": getColorFromToken(DEFAULT_BOUNDARIES_BORDER_COLOR),
          "line-width": [
            "case",
            ["==", ["get", "region_type"], "REGION"],
            REGION_BORDER_WIDTH,
            COUNTRY_BORDER_WIDTH
          ]
        }
      }
    )
  ] });
};
const MapInteractions = ({
  children,
  onContextLostError
}) => {
  const [contextLostError, setContextLostError] = useState(null);
  useHoverInteraction();
  useActiveInteraction();
  useWebGLContextError(() => {
    onContextLostError?.();
    setContextLostError(
      new Error(
        "Too many active maps. This is a browser limitation. Try reducing the number of map instances."
      )
    );
  });
  if (contextLostError) {
    throw contextLostError;
  }
  return children;
};
const MapContent = forwardRef((props, forwardedRef) => {
  const {
    initialViewState: initialViewStateProp,
    children,
    style,
    mapStyle = defaultStyle,
    onViewStateChange,
    onMapLoad,
    onContextLostError,
    ...remaining
  } = props;
  const loading = useMapLoading();
  const setZoomState = useSetStateZoom();
  const [savedInitialViewState, setSavedInitialViewState] = useState();
  const mapContainerRef = useRef(null);
  const mergedRefs = useMergeRefs([
    mapContainerRef,
    forwardedRef
  ]);
  const baseFeatureCollection = useMapBaseLayerFeatures();
  const dataLayersBBox = getDataLayersBoundingBox(
    children,
    baseFeatureCollection?.features
  );
  useTooltipEventListeners();
  const { toolbar: toolbarConfig, interactions: interactionsConfig } = useMapConfig();
  const layerIds = useContext(LayerIdsContext);
  const shouldRenderToolbar = !isUndefined(toolbarConfig) && !toolbarConfig.hidden || !isUndefined(interactionsConfig);
  const { height: graphHeight, width: graphWidth } = useGraphSize();
  const {
    handleMouseEnter,
    handleMouseMove,
    handleMouseLeave,
    handleMouseClick,
    handleZoom,
    handleDrag
  } = useOverlayEvents();
  const initialViewState = initialViewStateProp ?? dataLayersBBox;
  const { onLoadHandler, onRenderHandler } = useMapPerformance(mapContainerRef);
  const calculatedInitialViewState = toMapBoxInitialViewState(initialViewState);
  const { zoomToFitInProgress } = useMapZoomState();
  const handleZoomEnd = (event) => {
    if (savedInitialViewState) {
      const zoomState = generateNewZoomState(
        event,
        savedInitialViewState,
        zoomToFitInProgress
      );
      setZoomState(zoomState);
    }
  };
  const handleDragEnd = (event) => {
    if (savedInitialViewState) {
      const zoomState = generateNewZoomState(
        event,
        savedInitialViewState,
        zoomToFitInProgress
      );
      setZoomState(zoomState);
    }
  };
  const isZoomEnabled = toolbarConfig?.zoom?.enabled !== false;
  return /* @__PURE__ */ jsx(
    "div",
    {
      "data-testid": "map-container",
      ref: mergedRefs,
      onMouseEnter: handleMouseEnter,
      onMouseLeave: handleMouseLeave,
      children: /* @__PURE__ */ jsx(
        Map,
        {
          style: {
            width: graphWidth,
            height: `${graphHeight}px`,
            ...style
          },
          onMove: (evt) => {
            onViewStateChange?.(evt.viewState);
          },
          initialViewState: calculatedInitialViewState,
          mapStyle,
          dragRotate: false,
          fadeDuration: 0,
          renderWorldCopies: false,
          interactiveLayerIds: layerIds,
          maxBounds: [
            [MIN_LONGITUDE, MIN_LATITUDE],
            [MAX_LONGITUDE, MAX_LATITUDE]
          ],
          onMouseMove: (e) => handleMouseMove(e),
          onClick: (e) => handleMouseClick(e),
          onZoom: handleZoom,
          onZoomEnd: handleZoomEnd,
          onDrag: handleDrag,
          onDragEnd: handleDragEnd,
          boxZoom: isZoomEnabled,
          keyboard: isZoomEnabled,
          scrollZoom: isZoomEnabled,
          doubleClickZoom: isZoomEnabled,
          attributionControl: false,
          onRender: () => onRenderHandler && onRenderHandler(),
          onLoad: (e) => {
            onMapLoad();
            onLoadHandler && onLoadHandler();
            const initialCenter = e.target.getCenter();
            setSavedInitialViewState({
              latitude: initialCenter.lat,
              longitude: initialCenter.lng,
              zoom: e.target.getZoom()
            });
          },
          ref: useSetMapInitialConfiguration(),
          interactive: true,
          ...remaining,
          children: /* @__PURE__ */ jsx(MapDataBoundingBoxContext.Provider, { value: dataLayersBBox, children: /* @__PURE__ */ jsxs(
            MapViewImperativeHandler,
            {
              forwardedRef,
              containerRef: mapContainerRef,
              children: [
                /* @__PURE__ */ jsx(MapSource, {}),
                loading ? null : /* @__PURE__ */ jsxs(Fragment, { children: [
                  /* @__PURE__ */ jsx(MapInteractions, { onContextLostError, children }),
                  shouldRenderToolbar ? /* @__PURE__ */ jsx(MapToolbar, {}) : null
                ] })
              ]
            }
          ) })
        }
      )
    }
  );
});
MapContent.displayName = "MapContent";
export {
  MapContent
};
