// @dynatrace/strato-components/esm/tables/DataTable/features/Sorting/SortingArrowIndicators.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import { useIntl } from "react-intl";
import {
  ArrowSmallDownIcon,
  SorterSmallDoubleIcon
} from "@dynatrace/strato-icons";
import * as styles from "./Sorting.sty.js";
import { Text } from "../../../../typography/text/Text.js";
import { dataTableMessages } from "../../messages.js";
const SortingArrowIndicators = (props) => {
  const { isSorted, multiSortPriority, nextSortingOrder, rowDensity } = props;
  const intl = useIntl();
  const dataDtName = "ColumnHeaderSortIcon";
  let dataDtProperties = "state:clearSort";
  if (nextSortingOrder === "asc") {
    dataDtProperties = "state:ascending";
  } else if (nextSortingOrder === "desc") {
    dataDtProperties = "state:descending";
  }
  if (isSorted !== false) {
    const sortedDirectionIcon = /* @__PURE__ */ jsx(
      ArrowSmallDownIcon,
      {
        "data-sorticon": true,
        "data-dt-name": dataDtName,
        "data-dt-properties": dataDtProperties,
        style: { display: "inline" },
        className: clsx(
          styles.sortingArrow({ arrowUp: isSorted === "asc", rowDensity })
        )
      }
    );
    if (multiSortPriority !== void 0) {
      return /* @__PURE__ */ jsxs(Text, { as: "span", className: styles.multiSortIndicatorBubble, children: [
        sortedDirectionIcon,
        /* @__PURE__ */ jsx(
          Text,
          {
            as: "span",
            textStyle: "small-emphasized",
            "aria-hidden": "true",
            className: styles.sortOrderIndicator,
            children: multiSortPriority
          }
        )
      ] });
    }
    return sortedDirectionIcon;
  }
  return /* @__PURE__ */ jsx(
    SorterSmallDoubleIcon,
    {
      "data-sorticon": true,
      "aria-label": intl.formatMessage(dataTableMessages.sortableColumnIcon),
      className: styles.sortingArrow({ rowDensity }),
      "data-dt-name": dataDtName,
      "data-dt-properties": dataDtProperties
    }
  );
};
export {
  SortingArrowIndicators
};
