// @dynatrace-sdk/http-client/cjs/index.js
/**
 * @license
 * Copyright 2023-2026 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/http-client/src/index.ts
var src_exports = {};
__export(src_exports, {
  PlatformAbortController: () => AbortController,
  PlatformAbortSignal: () => AbortSignal,
  PlatformBaseError: () => BaseError,
  PlatformBinary: () => Binary,
  PlatformDataTypeError: () => DataTypeError,
  PlatformHttpClient: () => HttpClient,
  PlatformHttpClientAbortError: () => HttpClientAbortError,
  PlatformHttpClientRequestError: () => HttpClientRequestError,
  PlatformHttpClientResponse: () => HttpClientResponse,
  PlatformHttpClientResponseError: () => HttpClientResponseError,
  PlatformUnsupportedOperationError: () => UnsupportedOperationError,
  _OAuthHttpClient: () => OAuthHttpClient,
  httpClient: () => httpClient,
  isHttpClientAbortError: () => isHttpClientAbortError,
  isHttpClientNetworkError: () => isHttpClientNetworkError,
  isHttpClientRequestError: () => isHttpClientRequestError,
  isHttpClientResponseError: () => isHttpClientResponseError
});
module.exports = __toCommonJS(src_exports);

// packages/http-client/src/lib/types/http-client-request-error.ts
function isHttpClientRequestError(e) {
  return e?.name === "RequestError";
}

// packages/http-client/src/lib/types/http-client-abort-error.ts
function isHttpClientAbortError(e) {
  return e?.name === "AbortError";
}

// packages/http-client/src/lib/types/http-client-response-error.ts
function isHttpClientResponseError(e) {
  return e?.name === "ResponseError";
}

// packages/http-client/src/lib/types/http-client-network-error.ts
function isHttpClientNetworkError(e) {
  return e?.name === "NetworkError";
}

// packages/http-client/src/lib/platform/abort-controller.ts
var AbortController = globalThis.AbortController;

// packages/http-client/src/lib/platform/abort-signal.ts
var AbortSignal = globalThis.AbortSignal;

// packages/http-client/src/lib/platform/base-error.ts
var BaseError = class extends Error {
  constructor(message, cause) {
    super(message);
    this.cause = cause;
  }
};

// packages/http-client/src/lib/platform/unsupported-operation-error.ts
var UnsupportedOperationError = class extends BaseError {
  name = "UnsupportedOperationError";
  constructor(message, cause) {
    super(message, cause);
  }
};

// packages/http-client/src/lib/platform/binary.ts
function getImpl(data, type) {
  const response = new Response(data);
  switch (type) {
    case "text":
      return response.text();
    case "json":
      return response.json();
    case "array-buffer":
      return response.arrayBuffer();
    case "blob":
      return response.blob();
    case "readable-stream":
      return response.body;
  }
  throw new UnsupportedOperationError(`Unsupported binary type: ${type}.`);
}
var Binary = class _Binary {
  constructor(data) {
    this.data = data;
  }
  static fromText(text) {
    return new _Binary(new Blob([text]));
  }
  static fromJson(json) {
    return new _Binary(new Blob([JSON.stringify(json)]));
  }
  static from(data) {
    if (data instanceof Blob || data instanceof ArrayBuffer || data instanceof ReadableStream) {
      return new _Binary(data);
    } else {
      throw new UnsupportedOperationError("Unsupported binary type.");
    }
  }
  get(type) {
    return getImpl(this.data, type);
  }
};
function isBinary(body) {
  return typeof body.get === "function";
}

// packages/http-client/src/lib/platform/data-type-error.ts
var DataTypeError = class extends BaseError {
  name = "DataTypeError";
  constructor(message, cause) {
    super(message, cause);
  }
};

// packages/http-client/src/lib/platform/default-status-validator.ts
var defaultStatusValidator = (status) => 200 <= status && status < 300;

// packages/http-client/src/lib/platform/encode-request-body.ts
function encodeTextRequestBody(body) {
  return String(body);
}
function encodeJsonRequestBody(body) {
  return JSON.stringify(body);
}
async function encodeBinaryFormDataField(field) {
  if (field instanceof ArrayBuffer || field instanceof Blob) {
    return field;
  } else if (field instanceof ReadableStream) {
    return new Response(field).arrayBuffer();
  } else if (isBinary(field)) {
    return field.get("array-buffer");
  } else {
    throw new DataTypeError("Cannot encode form data field as binary type.");
  }
}
async function encodeFormDataRequestBody(body) {
  const formData = new FormData();
  for (const field of body) {
    switch (field.type) {
      case "text":
        formData.append(field.name, encodeTextRequestBody(field.value));
        break;
      case "json":
        formData.append(
          field.name,
          new Blob([encodeJsonRequestBody(field.value)], {
            type: field.contentType ?? "application/json"
          }),
          field.filename ?? "json"
        );
        break;
      case "binary": {
        if (File && field.value instanceof File) {
          const file = field.value;
          formData.append(
            field.name,
            new File([file], field.filename ?? file.name, {
              type: field.contentType ?? file.type ?? "application/octet-stream",
              lastModified: file.lastModified
            }),
            field.filename ?? file.name
          );
        } else {
          const blobType = "type" in field.value ? field.value.type : void 0;
          formData.append(
            field.name,
            new Blob([await encodeBinaryFormDataField(field.value)], {
              type: field.contentType ?? blobType ?? "application/octet-stream"
            }),
            field.filename ?? "binary"
          );
        }
        break;
      }
    }
  }
  return formData;
}
async function encodeBinaryRequestBody(body) {
  if (body instanceof Blob || body instanceof ArrayBuffer || body instanceof ReadableStream) {
    return body;
  } else if (isBinary(body)) {
    try {
      return await body.get("array-buffer");
    } catch (e) {
      throw new DataTypeError("Cannot encode body as binary type.", e);
    }
  } else {
    throw new DataTypeError("Cannot encode body as binary type.");
  }
}
async function encodeRequestBody(body, requestBodyType) {
  switch (requestBodyType) {
    case "text":
      return encodeTextRequestBody(body);
    case "json":
      return encodeJsonRequestBody(body);
    case "form-data":
      return await encodeFormDataRequestBody(body);
    case "binary":
      return await encodeBinaryRequestBody(body);
  }
  throw new DataTypeError(`Invalid body type: '${requestBodyType}'`);
}

// packages/http-client/src/lib/platform/apply-content-type-header.ts
function getContentTypeBy(bodyType) {
  switch (bodyType) {
    case "json":
      return "application/json";
    case "text":
      return "text/plain";
    case "binary":
      return "application/octet-stream";
  }
  throw new DataTypeError(`Invalid body type: '${bodyType}'`);
}
function normalizeHeaders(headers) {
  if (headers["Content-Type"]?.startsWith("multipart/")) {
    const { "Content-Type": contentTypeHeader, ...normalizedHeaders } = headers;
    return normalizedHeaders;
  }
  return headers;
}
function applyContentTypeHeader(headers, requestBodyType) {
  const normalizedHeaders = headers ? normalizeHeaders(headers) : void 0;
  const isContentTypePresent = Object.keys(normalizedHeaders ?? {}).some(
    (header) => header.toLowerCase() === "content-type"
  );
  if (!isContentTypePresent && requestBodyType !== "form-data") {
    return {
      "content-type": getContentTypeBy(requestBodyType),
      ...normalizedHeaders
    };
  }
  return normalizedHeaders;
}

// packages/http-client/src/lib/platform/http-client-abort-error.ts
var HttpClientAbortError = class extends Error {
  constructor(cause) {
    super(
      typeof cause?.message === "string" ? `Request aborted: ${cause.message}` : "Request aborted."
    );
    this.cause = cause;
  }
  name = "AbortError";
};
var throwIfAbortError = (e) => {
  if (e instanceof Error && e.name === "AbortError") {
    throw new HttpClientAbortError(e);
  }
};

// packages/http-client/src/lib/platform/http-client-request-error.ts
var HttpClientRequestError = class extends BaseError {
  constructor(cause) {
    super(
      typeof cause?.message === "string" ? `Request failed: ${cause.message}` : "Request failed."
    );
    this.cause = cause;
  }
  name = "RequestError";
};

// packages/http-client/src/lib/platform/decode-form-data.ts
function byteToChar(byte) {
  if (byte === void 0)
    return void 0;
  return String.fromCharCode(byte);
}
function isTextContentType(contentType) {
  return ["application/json", "text/plain"].includes(String(contentType).toLowerCase());
}
function removeEscaping(value) {
  if (!value)
    return value;
  return JSON.parse(`"${value}"`);
}
function parseContentDisposition(contentDisposition) {
  const fileName = removeEscaping(
    /form-data;.*filename="((?:[^"\\]|\\.)*)"/gm.exec(contentDisposition)?.[1]
  );
  const name = removeEscaping(
    /form-data;.*(?<!file)name="((?:[^"\\]|\\.)*)"/gm.exec(contentDisposition)?.[1]
  );
  return { name, fileName };
}
function parseFormDataBoundary(contentType) {
  return /multipart\/form-data;boundary=(.*)/gm.exec(contentType)?.[1];
}
var FormDataParser = class {
  constructor(sourceArray, boundary) {
    this.sourceArray = sourceArray;
    this.boundary = boundary;
  }
  currentChar = null;
  currentByte = null;
  offset = 0;
  utf8Decoder = new TextDecoder("utf-8");
  get previousByte() {
    return this.sourceArray[this.offset - 2];
  }
  get previousChar() {
    return byteToChar(this.previousByte);
  }
  next() {
    if (this.offset >= this.sourceArray.byteLength) {
      this.currentChar = null;
      this.currentByte = null;
      return false;
    }
    this.currentByte = this.sourceArray[this.offset];
    this.currentChar = String.fromCharCode(this.currentByte);
    this.offset++;
    return true;
  }
  charAt(offset) {
    return byteToChar(this.sourceArray[offset]) ?? null;
  }
  nextChar() {
    this.next();
    return this.currentChar;
  }
  nextByte() {
    this.next();
    return this.currentByte;
  }
  get nextTwoChars() {
    return String.fromCharCode(this.sourceArray[this.offset], this.sourceArray[this.offset + 1]);
  }
  findNextBoundaryOffset(startOffset = this.offset) {
    let boundaryIndex = 0;
    let boundaryMatch = false;
    let offset = startOffset;
    let currentChar = this.charAt(offset);
    while (!boundaryMatch) {
      if (currentChar === null) {
        return null;
      }
      if (currentChar === this.boundary[boundaryIndex]) {
        boundaryIndex++;
        if (boundaryIndex === this.boundary.length) {
          boundaryMatch = true;
        }
      } else {
        boundaryIndex = 0;
      }
      offset++;
      currentChar = this.charAt(offset);
    }
    return offset - this.boundary.length;
  }
  skipPastNextBoundary() {
    const nextBoundaryOffset = this.findNextBoundaryOffset();
    if (nextBoundaryOffset === null) {
      return false;
    }
    this.offset = nextBoundaryOffset + this.boundary.length;
    this.currentByte = this.sourceArray[this.offset];
    this.currentChar = String.fromCharCode(this.currentByte);
    const twoChars = this.nextTwoChars;
    if (twoChars == "--") {
      return false;
    } else if (twoChars == "\r\n") {
      this.offset++;
      this.next();
    }
    return true;
  }
  readLine() {
    const lineBuffer = [];
    do {
      const byte = this.nextByte();
      if (byte === null) {
        return null;
      }
      lineBuffer.push(byte);
    } while (this.currentChar !== null && this.currentChar !== "\n" && this.previousChar !== "\r");
    return Uint8Array.from(lineBuffer);
  }
  readSectionHeaders() {
    const headers = {};
    const headerStartOffset = this.offset;
    let isEndOfHeaders = false;
    while (!isEndOfHeaders) {
      const line = this.readLine();
      if (line === null) {
        throw TypeError("Unexpected end of response while parsing headers");
      }
      const headerString = this.utf8Decoder.decode(line);
      if (headerString === "\r\n") {
        isEndOfHeaders = true;
        break;
      }
      const [header, value] = headerString.split(":").map((part) => (part ?? "").trim());
      headers[header.toLowerCase()] = value;
    }
    const headersLength = this.offset - headerStartOffset;
    return {
      headersLength,
      headers
    };
  }
  extractData() {
    const nextBoundaryOffset = this.findNextBoundaryOffset() ?? this.sourceArray.length;
    const lastTwoBytes = String.fromCharCode(
      this.sourceArray[nextBoundaryOffset - 2],
      this.sourceArray[nextBoundaryOffset - 1]
    );
    const endOfDataSection = lastTwoBytes === "\r\n" ? nextBoundaryOffset - 2 : nextBoundaryOffset;
    const data = this.sourceArray.slice(this.offset, endOfDataSection);
    return {
      payloadStartOffset: this.offset,
      payloadEndOffset: endOfDataSection,
      data
    };
  }
  parse() {
    const sections = [];
    while (this.skipPastNextBoundary()) {
      const headersStartOffset = this.offset;
      const { headers } = this.readSectionHeaders();
      const { data, payloadStartOffset, payloadEndOffset } = this.extractData();
      const isText = isTextContentType(headers["content-type"]);
      const { name, fileName } = parseContentDisposition(headers["content-disposition"]);
      sections.push({
        name,
        text: isText ? this.utf8Decoder.decode(data) : void 0,
        file: !isText ? new File([data], fileName ?? "", {
          type: headers["content-type"] ?? "application/octet-stream"
        }) : void 0,
        rawData: data,
        headers,
        headersStartOffset,
        payloadStartOffset,
        payloadEndOffset
      });
    }
    return sections;
  }
};
function isAppleBrowser() {
  return /apple/i.test(navigator.vendor);
}
function parseMultipartFormDataResponse(response, boundary) {
  return new FormDataParser(response, boundary).parse();
}

// packages/http-client/src/lib/platform/decode-response-body.ts
function emptyReadableStream() {
  return new ReadableStream({
    start(controller) {
      controller.close();
    }
  });
}
async function decodeTextResponseBody(response) {
  return response.body !== null ? response.text() : "";
}
async function decodeJsonResponseBody(response) {
  return response.body !== null ? response.json() : null;
}
async function decodeBinaryResponseBody(response) {
  return Binary.from(decodeReadableStreamResponseBody(response));
}
async function decodeFormDataResponseBody(response) {
  if (response.body === null) {
    return [];
  } else {
    const result = [];
    if (isAppleBrowser()) {
      const responseClone = await response.clone();
      const responseBinaryData = new Uint8Array(await responseClone.arrayBuffer());
      const boundary = parseFormDataBoundary(responseClone.headers.get("content-type"));
      const manuallyParsedFormData = parseMultipartFormDataResponse(
        responseBinaryData,
        `--${boundary}`
      );
      const automaticallyParsedFormData = await response.formData();
      automaticallyParsedFormData.forEach((value, key) => {
        const manuallyParsedSection = manuallyParsedFormData.find(({ name }) => name === key);
        if (typeof value === "string" && manuallyParsedSection?.file !== void 0) {
          result.push({
            type: "binary",
            name: key,
            contentType: manuallyParsedSection.headers["content-type"],
            filename: manuallyParsedSection.file.name,
            value: Binary.from(manuallyParsedSection.file)
          });
        } else if (value instanceof File) {
          result.push({
            type: "binary",
            name: key,
            contentType: value.type,
            filename: value.name,
            value: Binary.from(value)
          });
        } else {
          result.push({
            type: "text",
            name: key,
            value
          });
        }
      });
    } else {
      const formData = await response.formData();
      formData.forEach((value, key) => {
        if (value instanceof File) {
          result.push({
            type: "binary",
            name: key,
            contentType: value.type,
            filename: value.name,
            value: Binary.from(value)
          });
        } else {
          result.push({
            type: "text",
            name: key,
            value
          });
        }
      });
    }
    return result;
  }
}
async function decodeArrayBufferResponseBody(response) {
  return response !== null ? response.arrayBuffer() : new ArrayBuffer(0);
}
async function decodeBlobResponseBody(response) {
  return response !== null ? response.blob() : new Blob();
}
function decodeReadableStreamResponseBody(response) {
  return response.body !== null ? response.body : emptyReadableStream();
}
function throwResponseBodyTypeError(response, responseBodyType, cause) {
  throw new DataTypeError(
    `Response body does not conform to the specified response body type: '${responseBodyType}'. The content type of the response is '${response.headers.get(
      "Content-Type"
    )}'. Response status is '${response.status}'. Response source is '${response.headers.get(
      "dynatrace-response-source"
    )}'`,
    cause
  );
}
async function decodeAsyncResponseBody(response, responseBodyType) {
  try {
    switch (responseBodyType) {
      case "text":
        return await decodeTextResponseBody(response);
      case "json":
        return await decodeJsonResponseBody(response);
      case "binary":
        return await decodeBinaryResponseBody(response);
      case "form-data":
        return await decodeFormDataResponseBody(response);
      case "array-buffer":
        return await decodeArrayBufferResponseBody(response);
      case "blob":
        return await decodeBlobResponseBody(response);
    }
  } catch (e) {
    throwIfAbortError(e);
    throwResponseBodyTypeError(response, responseBodyType, e);
  }
}
function decodeSyncResponseBody(response, responseBodyType) {
  try {
    return decodeReadableStreamResponseBody(response);
  } catch (e) {
    throwIfAbortError(e);
    throwResponseBodyTypeError(response, responseBodyType, e);
  }
}
function decodeResponseBodyImpl(response, responseBodyType) {
  switch (responseBodyType) {
    case "text":
    case "json":
    case "binary":
    case "form-data":
    case "array-buffer":
    case "blob":
      return decodeAsyncResponseBody(response, responseBodyType);
    case "readable-stream":
      return decodeSyncResponseBody(response, responseBodyType);
  }
  switch (responseBodyType) {
    case "buffer":
    case "stream":
      throw new UnsupportedOperationError(`Unsupported response body type: '${responseBodyType}'`);
    default:
      throw new DataTypeError(`Invalid body type: '${responseBodyType}'`);
  }
}
function decodeResponseBody(response, responseBodyType) {
  return decodeResponseBodyImpl(response, responseBodyType);
}

// packages/http-client/src/lib/platform/http-client-response.ts
var HttpClientResponse = class {
  constructor(response, data) {
    this.response = response;
    this.requestUrl = data?.requestUrl ?? response.url;
  }
  requestUrl;
  get url() {
    return this.response.url;
  }
  get status() {
    return this.response.status;
  }
  get statusText() {
    return this.response.statusText;
  }
  get headers() {
    const headers = {};
    this.response.headers.forEach((value, key) => {
      headers[key] = value;
    });
    return headers;
  }
  body(responseBodyType = "json") {
    return decodeResponseBody(this.response.clone(), responseBodyType);
  }
};

// packages/http-client/src/lib/platform/http-client-response-error-serializer.ts
var import_error_handlers = require("@dynatrace-sdk/error-handlers");
function getMessageFromObject(obj) {
  if (typeof obj === "object" && obj !== null && typeof obj.message === "string") {
    return obj.message;
  }
}
async function parseResponse(response) {
  let consumedBody;
  try {
    consumedBody = await response.body("text");
  } catch (e) {
    const bodyError = `Response body is not available. ${e?.message ?? "Unknown error occurred."}`;
    return {
      message: bodyError.substring(0, 100),
      body: bodyError
    };
  }
  try {
    const jsonBody = JSON.parse(consumedBody);
    if (typeof jsonBody === "object" && jsonBody !== null) {
      return {
        message: getMessageFromObject(jsonBody.error) || getMessageFromObject(jsonBody) || consumedBody.substring(0, 100),
        body: jsonBody,
        errorCode: jsonBody.error?.details?.errorCode,
        errorRef: jsonBody.error?.details?.errorRef,
        traceId: jsonBody.error?.details?.traceId
      };
    } else {
      return {
        message: consumedBody.substring(0, 100),
        body: consumedBody
      };
    }
  } catch (e) {
    return {
      message: consumedBody.substring(0, 100),
      body: consumedBody
    };
  }
}
var httpClientResponseErrorSerializer = async (error) => {
  if (error instanceof HttpClientResponseError) {
    const status = error.response.status;
    const { message, body, errorRef, traceId, errorCode } = await parseResponse(error.response);
    const optionalErrorRef = errorRef ? { errorRef } : {};
    const optionalTraceId = traceId ? { traceId } : {};
    const stringifiedBody = JSON.stringify(body);
    const bodySample = stringifiedBody.length > 100 ? `${stringifiedBody.substring(0, 100)}\u2026` : stringifiedBody;
    return {
      name: error.name,
      status,
      message,
      stack: error.stack,
      type: import_error_handlers.ErrorType.HTTP,
      details: {
        url: error.response.url,
        requestUrl: error.response.requestUrl,
        method: error.requestMethod,
        errorCode,
        responseContentType: getHeaderValue(error.response, "content-type"),
        responseContentLength: getHeaderValue(error.response, "content-length"),
        responseDtSource: getHeaderValue(error.response, "dynatrace-response-source"),
        responseBodySample: bodySample
      },
      ...optionalErrorRef,
      ...optionalTraceId
    };
  }
};
var getHeaderValue = (response, header) => {
  const { headers = {} } = response;
  const headerKey = Object.keys(headers).find((key) => key.toLowerCase() === header.toLowerCase());
  return headerKey ? headers[headerKey] : void 0;
};
var errorSerializerAdded = false;
function addHttpClientResponseErrorSerializer() {
  if (!errorSerializerAdded && (0, import_error_handlers.isGlobalErrorSerializerSupported)()) {
    errorSerializerAdded = true;
    (0, import_error_handlers.addGlobalErrorSerializer)(httpClientResponseErrorSerializer);
  }
}

// packages/http-client/src/lib/platform/http-client-response-error.ts
var HttpClientResponseError = class extends BaseError {
  name = "ResponseError";
  response;
  requestMethod;
  constructor(response, data) {
    addHttpClientResponseErrorSerializer();
    super(`HTTP Error ${response.status}: ${response.statusText}`);
    this.response = new HttpClientResponse(response, {
      requestUrl: data?.requestUrl ?? response.url
    });
    this.requestMethod = data?.requestMethod;
  }
};

// packages/http-client/src/lib/platform/http-client.ts
var send = async (info, init) => {
  try {
    return await fetch(info, init);
  } catch (e) {
    throwIfAbortError(e);
    throw new HttpClientRequestError(e);
  }
};
var HttpClient = class {
  _baseUrl;
  _defaultHeaders;
  constructor(options) {
    this._baseUrl = options?.baseUrl;
    this._defaultHeaders = options?.defaultHeaders;
  }
  /**
   * Sets HTTP Client Instance baseUrl param to the incoming value `baseUrl`.
   * @param baseUrl  Base URL used in the request path.
   */
  _setBaseURL(baseUrl) {
    this._baseUrl = baseUrl;
  }
  /**
   * Sets HTTP Client Instance defaultHeaders param to the incoming value `defaultHeaders`.
   * @param defaultHeaders  HTTP Headers to be used in outgoing requests.
   */
  _setDefaultHeaders(defaultHeaders) {
    this._defaultHeaders = defaultHeaders;
  }
  async send(options) {
    const { statusValidator = defaultStatusValidator } = options;
    const requestBodyType = options.requestBodyType ?? "json";
    const body = options.body !== void 0 ? await encodeRequestBody(options.body, requestBodyType) : void 0;
    const url = this._baseUrl ? new URL(options.url, this._baseUrl).toString() : options.url;
    const requestHeaders = { ...this._defaultHeaders, ...options?.headers };
    const response = await send(url, {
      method: options.method,
      headers: applyContentTypeHeader(requestHeaders, requestBodyType),
      signal: options.abortSignal,
      body
    });
    const ok = statusValidator(response.status);
    if (ok) {
      return new HttpClientResponse(response, { requestUrl: url });
    } else {
      throw new HttpClientResponseError(response, {
        requestUrl: url,
        requestMethod: options.method
      });
    }
  }
};
var httpClient = new HttpClient();

// packages/http-client/src/lib/platform/oauth-http-client.ts
var TOKEN_EXPIRATION_SAFETY_MARGIN_SECONDS = 1;
var OAuthToken = class {
  _scope;
  _tokenType;
  _expirationDate;
  _accessToken;
  constructor(oauthTokenResponse) {
    this._scope = oauthTokenResponse.scope;
    this._tokenType = oauthTokenResponse.token_type;
    this._expirationDate = ((expiresIn) => {
      const date = /* @__PURE__ */ new Date();
      date.setSeconds(date.getSeconds() + expiresIn - TOKEN_EXPIRATION_SAFETY_MARGIN_SECONDS);
      return date;
    })(oauthTokenResponse.expires_in);
    this._accessToken = oauthTokenResponse.access_token;
  }
  get scope() {
    return this._scope;
  }
  get tokenType() {
    return this._tokenType;
  }
  isExpired() {
    return this._expirationDate <= /* @__PURE__ */ new Date();
  }
  get accessToken() {
    return this._accessToken;
  }
};
var send2 = async (info, init) => {
  try {
    return await fetch(info, init);
  } catch (e) {
    throwIfAbortError(e);
    throw new HttpClientRequestError(e);
  }
};
var OAuthHttpClient = class {
  constructor(config) {
    this.config = config;
  }
  oauthToken;
  async send(options) {
    const { statusValidator = defaultStatusValidator } = options;
    if (!this.oauthToken || this.oauthToken.isExpired()) {
      this.oauthToken = await this.requestToken();
    }
    const requestBodyType = options.requestBodyType ?? "json";
    const url = new URL(options.url, this.config.environmentUrl);
    const body = options.body !== void 0 ? await encodeRequestBody(options.body, requestBodyType) : void 0;
    const headers = {
      ...applyContentTypeHeader(options.headers, requestBodyType),
      Authorization: `${this.oauthToken.tokenType} ${this.oauthToken.accessToken}`
    };
    const requestUrl = url.toString();
    const response = await send2(requestUrl, {
      method: options.method,
      signal: options.abortSignal,
      body,
      headers
    });
    const ok = statusValidator(response.status);
    if (ok) {
      return new HttpClientResponse(response, { requestUrl });
    } else {
      throw new HttpClientResponseError(response, {
        requestUrl,
        requestMethod: options.method
      });
    }
  }
  async requestToken() {
    const res = await send2(this.config.authUrl, {
      method: "POST",
      headers: {
        "Content-Type": "application/x-www-form-urlencoded"
      },
      body: new URLSearchParams({
        grant_type: "client_credentials",
        client_id: this.config.clientId,
        client_secret: this.config.secret,
        scope: this.config.scopes.join(" ")
      })
    });
    const data = await res.json();
    return new OAuthToken(data);
  }
};
