// @dynatrace/strato-components/esm/charts/core/components/chart-layout/types/constants.js
import { defaultRatio } from "../../../constants.js";
const DEFAULT_POSITION = "auto";
const DEFAULT_RATIO_BOUNDARIES = {
  minWidth: 0.1,
  minHeight: 0.1,
  maxWidth: 0.8,
  maxHeight: 0.7
};
const DEFAULT_LAYOUT_HEIGHT = "300";
const DEFAULT_LAYOUT_WIDTH = "100%";
const DEFAULT_LEGEND_CONFIG = {
  position: DEFAULT_POSITION,
  ratioBoundaries: DEFAULT_RATIO_BOUNDARIES,
  ratio: defaultRatio,
  overrideRatio: void 0
};
const LOADING_STATE_OPACITY = 40;
const DEFAULT_RESIZER_SIZE = 15;
const MAX_AUTO_RATIO_SIZE = 0.3;
const MIN_LEGEND_HEIGHT = 30;
const MIN_LEGEND_WIDTH = 70;
const MIN_PLOT_SIZE = 100;
const TEMPLATE_AREA_FOR_GRAPH = '"graphArea"';
const TEMPLATE_AREA_FOR_RESIZABLE_CHART_LAYOUT = '"graphArea horizontalResize verticalLegend" "verticalResize . ." "horizontalLegend horizontalLegend horizontalLegend"';
const TEMPLATE_AREA_FOR_FIXED_CHART_LAYOUT = '"graphArea verticalLegend" "horizontalLegend horizontalLegend"';
export {
  DEFAULT_LAYOUT_HEIGHT,
  DEFAULT_LAYOUT_WIDTH,
  DEFAULT_LEGEND_CONFIG,
  DEFAULT_POSITION,
  DEFAULT_RATIO_BOUNDARIES,
  DEFAULT_RESIZER_SIZE,
  LOADING_STATE_OPACITY,
  MAX_AUTO_RATIO_SIZE,
  MIN_LEGEND_HEIGHT,
  MIN_LEGEND_WIDTH,
  MIN_PLOT_SIZE,
  TEMPLATE_AREA_FOR_FIXED_CHART_LAYOUT,
  TEMPLATE_AREA_FOR_GRAPH,
  TEMPLATE_AREA_FOR_RESIZABLE_CHART_LAYOUT
};
