// @dynatrace/strato-components/esm/charts/core/components/overlay-tooltip/components/global-actions/Action.js
import { jsx, jsxs } from "react/jsx-runtime";
import { actionButton } from "./actions.sty.js";
import { Button } from "../../../../../../buttons/button/Button.js";
import { useTooltipReducer } from "../../useTooltipReducer.js";
const Action = (props) => {
  const { label, icon, trailingIcon, onSelect, disabled } = props;
  const dispatch = useTooltipReducer();
  const handleKeyDown = (event) => {
    if (event.key === "Enter" || event.key === " ") {
      event.preventDefault();
      event.stopPropagation();
      onSelect(event);
    }
  };
  return /* @__PURE__ */ jsxs(
    Button,
    {
      disabled,
      variant: "default",
      className: actionButton,
      onClick: onSelect,
      onKeyDown: handleKeyDown,
      onPointerEnter: dispatch({
        type: "SET_ACTIVE_SERIES",
        payload: { activeSeries: "global-action" }
      }),
      width: "full",
      textAlign: "start",
      children: [
        icon && /* @__PURE__ */ jsx(Button.Prefix, { children: icon }),
        /* @__PURE__ */ jsx(Button.Label, { children: label }),
        trailingIcon && /* @__PURE__ */ jsx(Button.Suffix, { children: trailingIcon })
      ]
    }
  );
};
export {
  Action
};
