// @dynatrace/strato-components/filters/filter-field/utils/validation-utils/normalize-validator-map.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var normalize_validator_map_exports = {};
__export(normalize_validator_map_exports, {
  KEY_GROUP_ID: () => KEY_GROUP_ID,
  toInternalValidatorMap: () => toInternalValidatorMap
});
module.exports = __toCommonJS(normalize_validator_map_exports);
var import_type_guards = require("./type-guards.js");
var import_suggestion_group_utils = require("../suggestions-utils/suggestion-group-utils.js");
const KEY_GROUP_ID = "strato-filter-field-key-suggestions";
function toInternalValidatorMap(validatorMap) {
  if (!validatorMap) {
    return void 0;
  }
  const byKey = /* @__PURE__ */ new Map();
  const valuesMap = /* @__PURE__ */ new Map();
  const functionsMap = /* @__PURE__ */ new Map();
  const fallbackKeysList = [];
  let groups;
  const { keyPredicates, exhaustive, comparisonOperators } = validatorMap;
  if (keyPredicates !== void 0) {
    if (Array.isArray(keyPredicates)) {
      groups = processKeyPredicatesArray(
        keyPredicates,
        byKey,
        valuesMap,
        functionsMap,
        fallbackKeysList
      );
    } else {
      processKeyPredicatesObject(
        keyPredicates,
        byKey,
        valuesMap,
        functionsMap,
        fallbackKeysList
      );
    }
  }
  const filterStatementKeysMap = valuesMap.size > 0 || functionsMap.size > 0 ? { values: valuesMap, functions: functionsMap } : void 0;
  const resolvedGroups = groups && groups.length > 0 ? groups : void 0;
  const keyToGroup = /* @__PURE__ */ new Map();
  if (resolvedGroups) {
    for (const group of resolvedGroups) {
      for (const key of group.keys) {
        keyToGroup.set(key, group);
      }
    }
  } else {
    const implicitGroup = {
      id: `${KEY_GROUP_ID}-implicit-0`,
      keys: [...byKey.keys()]
    };
    for (const key of byKey.keys()) {
      keyToGroup.set(key, implicitGroup);
    }
  }
  return {
    normalizedKeys: {
      byKey,
      groups: resolvedGroups,
      keyToGroup
    },
    exhaustive,
    comparisonOperators,
    sortSuggestions: validatorMap.sortSuggestions,
    fallbackKeys: fallbackKeysList.length > 0 ? fallbackKeysList : void 0,
    filterStatementKeysMap
  };
}
function processKeyPredicatesArray(keyPredicates, byKey, valuesMap, functionsMap, fallbackKeysList) {
  const groups = [];
  let pendingKeys = [];
  function processPendingKeys() {
    if (pendingKeys.length === 0) {
      return;
    }
    groups.push({
      id: `${KEY_GROUP_ID}-implicit-${groups.length}`,
      keys: pendingKeys
    });
    pendingKeys = [];
  }
  for (let index = 0; index < keyPredicates.length; index++) {
    const item = keyPredicates[index];
    if (typeof item === "string") {
      byKey.set(item, {});
      pendingKeys.push(item);
      continue;
    }
    if ((0, import_type_guards.isFilterFieldKeySuggestionConfig)(item)) {
      byKey.set(item.key, item);
      collectFallbackAndStatements(
        item.key,
        item,
        fallbackKeysList,
        valuesMap,
        functionsMap
      );
      pendingKeys.push(item.key);
      continue;
    }
    if ((0, import_type_guards.isFilterFieldKeySuggestionGroupConfig)(item)) {
      processPendingKeys();
      const { label, "aria-label": ariaLabel } = item;
      const groupId = (0, import_suggestion_group_utils.resolveGroupId)(item, `key-group-${index}`);
      const groupKeys = [];
      for (const suggestion of item.suggestions) {
        if (typeof suggestion === "string") {
          byKey.set(suggestion, {});
          groupKeys.push(suggestion);
        } else if ((0, import_type_guards.isFilterFieldKeySuggestionConfig)(suggestion)) {
          byKey.set(suggestion.key, suggestion);
          groupKeys.push(suggestion.key);
          collectFallbackAndStatements(
            suggestion.key,
            suggestion,
            fallbackKeysList,
            valuesMap,
            functionsMap
          );
        }
      }
      if (groupKeys.length > 0) {
        groups.push({
          id: groupId,
          label,
          "aria-label": ariaLabel,
          keys: groupKeys
        });
      }
    }
  }
  processPendingKeys();
  return groups;
}
function processKeyPredicatesObject(keyPredicates, byKey, valuesMap, functionsMap, fallbackKeysList) {
  for (const [key, rawPredicate] of Object.entries(keyPredicates)) {
    let config;
    if ((0, import_type_guards.isFilterFieldValidatorMapTypeSpecification)(rawPredicate)) {
      config = { valuePredicate: rawPredicate };
    } else {
      config = {
        comparisonOperators: rawPredicate.comparisonOperators,
        valuePredicate: rawPredicate.valuePredicate,
        fallbackKey: rawPredicate.fallbackKey,
        suggestStatementOnValueMatch: rawPredicate.suggestStatementOnValueMatch,
        displayValue: rawPredicate.displayValue,
        details: rawPredicate.details
      };
    }
    byKey.set(key, config);
    collectFallbackAndStatements(
      key,
      config,
      fallbackKeysList,
      valuesMap,
      functionsMap
    );
  }
}
function collectFallbackAndStatements(key, config, fallbackKeysList, valuesMap, functionsMap) {
  if (config.fallbackKey) {
    fallbackKeysList.push({
      value: key,
      comparisonOperators: Array.isArray(config.fallbackKey) ? config.fallbackKey : void 0
    });
  }
  if (config.suggestStatementOnValueMatch) {
    if (typeof config.suggestStatementOnValueMatch === "function") {
      functionsMap.set(key, config.suggestStatementOnValueMatch);
    } else if (config.valuePredicate !== void 0 && typeof config.valuePredicate !== "function") {
      extractLiteralValuesToMap(config.valuePredicate, key, valuesMap);
    }
  }
}
function extractLiteralValuesToMap(valuePredicate, key, valuesMap) {
  if (Array.isArray(valuePredicate)) {
    for (const item of valuePredicate) {
      if ((0, import_type_guards.isFilterFieldValueSuggestionGroupConfig)(item)) {
        for (const suggestion of item.suggestions) {
          if (!(0, import_type_guards.isFilterFieldValidatorMapTypeSpecification)(suggestion)) {
            addToValuesMap(valuesMap, key, suggestion);
          }
        }
      } else if (!(0, import_type_guards.isFilterFieldValidatorMapTypeSpecification)(item)) {
        addToValuesMap(valuesMap, key, item);
      }
    }
  } else if ((0, import_type_guards.isFilterFieldValueSuggestionGroupConfig)(valuePredicate)) {
    for (const suggestion of valuePredicate.suggestions) {
      if (!(0, import_type_guards.isFilterFieldValidatorMapTypeSpecification)(suggestion)) {
        addToValuesMap(valuesMap, key, suggestion);
      }
    }
  } else if (!(0, import_type_guards.isFilterFieldValidatorMapTypeSpecification)(valuePredicate)) {
    addToValuesMap(valuesMap, key, valuePredicate);
  }
}
function addToValuesMap(valuesMap, key, value) {
  let stringValue;
  if ((0, import_type_guards.isFilterFieldValidatorMapDurationSpecification)(value)) {
    stringValue = `${value.value}${value.unit}`;
  } else if ((0, import_type_guards.isFilterFieldRichValuePredicate)(value)) {
    stringValue = `${value.value}`;
  } else {
    stringValue = `${value}`;
  }
  const existingKeys = valuesMap.get(stringValue) ?? [];
  if (!existingKeys.includes(key)) {
    valuesMap.set(stringValue, [...existingKeys, key]);
  }
}
