// @dynatrace/strato-components/filters/filter-field/utils/validation-utils/operator-specification.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var operator_specification_exports = {};
__export(operator_specification_exports, {
  FILTERFIELD_DEFAULT_OPERATORS: () => FILTERFIELD_DEFAULT_OPERATORS,
  getAllowedOperatorsForType: () => getAllowedOperatorsForType,
  getOperatorName: () => getOperatorName,
  getOperatorShortName: () => getOperatorShortName
});
module.exports = __toCommonJS(operator_specification_exports);
const OperatorsWithNames = {
  "=": "equals",
  "!=": "not-equals",
  "<": "less-than",
  "<=": "less-or-equal",
  ">": "greater-than",
  ">=": "greater-or-equal",
  "= *": "exists",
  // first match
  "=*": "exists",
  "!= *": "not-exists",
  // first match
  "!=*": "not-exists",
  "~": "matches-phrase",
  "!~": "not-matches-phrase",
  "*~": "search",
  "* ~": "search",
  in: "in",
  "not in": "not in",
  contains: "contains",
  "starts-with": "starts-with",
  "ends-with": "ends-with",
  "not-contains": "not-contains",
  "not-starts-with": "not-starts-with",
  "not-ends-with": "not-ends-with"
};
const getOperatorName = (operator) => OperatorsWithNames[operator] ?? "";
const getOperatorShortName = (comparisonOperator) => {
  for (const [key, value] of Object.entries(OperatorsWithNames)) {
    if (value === comparisonOperator) {
      return key;
    }
  }
  return void 0;
};
const baseOperators = [
  "equals",
  "not-equals",
  "exists",
  "not-exists"
];
const stringOperators = [
  "matches-phrase",
  "not-matches-phrase",
  "contains",
  "starts-with",
  "ends-with",
  "not-contains",
  "not-starts-with",
  "not-ends-with"
];
const listOperators = ["in", "not in"];
const numericOperators = [
  "less-than",
  "less-or-equal",
  "greater-than",
  "greater-or-equal"
];
const getAllowedOperatorsForType = (type) => {
  switch (type) {
    case "String":
      return [...baseOperators, ...listOperators, ...stringOperators];
    case "Number":
      return [...baseOperators, ...numericOperators, ...listOperators];
    case "Boolean":
      return [...baseOperators];
    case "Timestamp":
    case "IPAddress":
    case "Duration":
      return [...baseOperators, ...numericOperators, ...listOperators];
    case "UID":
    case "SmartscapeId":
      return [...baseOperators, ...listOperators];
    case "JSONPath":
    case "Any":
    default:
      return [
        ...baseOperators,
        ...numericOperators,
        ...listOperators,
        ...stringOperators
      ];
  }
};
const FILTERFIELD_DEFAULT_OPERATORS = [
  "equals",
  "not-equals",
  "greater-than",
  "greater-or-equal",
  "less-than",
  "less-or-equal",
  "in",
  "not in",
  "starts-with",
  "contains",
  "ends-with",
  "not-starts-with",
  "not-contains",
  "not-ends-with",
  "exists",
  "not-exists"
];
