// @dynatrace/strato-components/filters/filter-field/utils/validation-utils/type-guards.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var type_guards_exports = {};
__export(type_guards_exports, {
  isFilterFieldKeySuggestionConfig: () => isFilterFieldKeySuggestionConfig,
  isFilterFieldKeySuggestionGroupConfig: () => isFilterFieldKeySuggestionGroupConfig,
  isFilterFieldRichValuePredicate: () => isFilterFieldRichValuePredicate,
  isFilterFieldValidatorMapDurationSpecification: () => isFilterFieldValidatorMapDurationSpecification,
  isFilterFieldValidatorMapTypeSpecification: () => isFilterFieldValidatorMapTypeSpecification,
  isFilterFieldValueSuggestionGroupConfig: () => isFilterFieldValueSuggestionGroupConfig
});
module.exports = __toCommonJS(type_guards_exports);
function isFilterFieldValidatorMapTypeSpecification(valueSpecification) {
  return typeof valueSpecification === "object" && valueSpecification !== null && "type" in valueSpecification && !("value" in valueSpecification);
}
function isFilterFieldValidatorMapDurationSpecification(valueSpecification) {
  const castedValueSpecification = valueSpecification;
  return typeof valueSpecification === "object" && castedValueSpecification.value !== void 0 && castedValueSpecification.unit !== void 0;
}
function isFilterFieldRichValuePredicate(valueSpecification) {
  const castedValueSpecification = valueSpecification;
  return typeof valueSpecification === "object" && castedValueSpecification !== null && "value" in castedValueSpecification && !isFilterFieldValidatorMapTypeSpecification(valueSpecification) && ("displayValue" in castedValueSpecification || "details" in castedValueSpecification);
}
function isFilterFieldKeySuggestionConfig(value) {
  return typeof value === "object" && value !== null && "key" in value && typeof value.key === "string";
}
function isFilterFieldKeySuggestionGroupConfig(value) {
  return typeof value === "object" && value !== null && "suggestions" in value && Array.isArray(value.suggestions) && !("key" in value);
}
function isFilterFieldValueSuggestionGroupConfig(value) {
  return typeof value === "object" && value !== null && "suggestions" in value && Array.isArray(value.suggestions);
}
