// @dynatrace/strato-components/filters/filter-field/utils/validation-utils/validation-helper-functions.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var validation_helper_functions_exports = {};
__export(validation_helper_functions_exports, {
  getColoredSearchOperator: () => getColoredSearchOperator,
  getValidOperators: () => getValidOperators,
  getValidOperatorsColored: () => getValidOperatorsColored,
  getValidOperatorsShort: () => getValidOperatorsShort,
  getValuePredicateTextValue: () => getValuePredicateTextValue,
  isValidCustomType: () => isValidCustomType,
  isValidOperator: () => isValidOperator,
  isValidValue: () => isValidValue,
  isWildCardOperator: () => isWildCardOperator
});
module.exports = __toCommonJS(validation_helper_functions_exports);
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
var import_operator_specification = require("./operator-specification.js");
var import_type_guards = require("./type-guards.js");
var import_type_validators = require("./type-validators.js");
var import_is_node_environment = require("../../../../core/utils/is-node-environment.js");
var import_suggestion_type = require("../suggestions-utils/suggestion-type.js");
function isWildCardOperator(currentStatement) {
  const wildCardOperators = [
    "starts-with",
    "ends-with",
    "contains",
    "not-starts-with",
    "not-ends-with",
    "not-contains"
  ];
  return typeof currentStatement?.comparisonOperator?.value === "string" && wildCardOperators.includes(currentStatement?.comparisonOperator?.value);
}
function isValidIPAddressType(valueType, textValue, validTypes, parserConfig) {
  if (!parserConfig?.ipAddressConversion && validTypes.includes("IPAddress")) {
    if ((0, import_is_node_environment.isDev)()) {
      console.warn(
        "IPAddress is configured as a type in the validatorMap of the FilterField. To get full support for IPAddress types please also enable IPAddress conversion in the parserConfig."
      );
    }
    return valueType === "Identifier" && (0, import_type_validators.isValidIPAddress)(textValue);
  } else if (validTypes.includes("IPAddress")) {
    return valueType === "IPAddress";
  }
  return false;
}
function isValidTimestampType(valueType, textValue, validTypes, parserConfig) {
  if (!parserConfig?.timestampConversion && validTypes.includes("Timestamp")) {
    if ((0, import_is_node_environment.isDev)()) {
      console.warn(
        "Timestamp is configured as a type in the validatorMap of the FilterField. To get full support for Timestamp types please also enable Timestamp conversion in the parserConfig."
      );
    }
    return (valueType === "Identifier" || valueType === "String") && (0, import_type_validators.isValidTimestamp)(textValue);
  } else if (validTypes.includes("Timestamp")) {
    return valueType === "Timestamp";
  }
  return false;
}
function isValidUIDType(valueType, textValue, validTypes, parserConfig) {
  if (!parserConfig?.uidConversion && validTypes.includes("UID")) {
    if ((0, import_is_node_environment.isDev)()) {
      console.warn(
        "UID is configured as a type in the validatorMap of the FilterField. To get full support for UID types please also enable UID conversion in the parserConfig."
      );
    }
    return valueType === "Identifier" && (0, import_type_validators.isValidUID)(textValue);
  } else if (validTypes.includes("UID")) {
    return valueType === "UID";
  }
  return false;
}
function isValidSmartscapeIdType(valueType, textValue, validTypes, parserConfig) {
  if (
    // eslint-disable-next-line @typescript-eslint/no-deprecated
    !parserConfig?.smartscapeIdConversion && validTypes.includes("SmartscapeId")
  ) {
    if ((0, import_is_node_environment.isDev)()) {
      console.warn(
        "SmartscapeId is configured as a type in the validatorMap of the FilterField. To get full support for SmartscapeId types please also enable SmartscapeId conversion in the parserConfig."
      );
    }
    return valueType === "Identifier" && (0, import_type_validators.isValidSmartscapeId)(textValue);
  } else if (validTypes.includes("SmartscapeId")) {
    return valueType === "SmartscapeId";
  }
  return false;
}
function checkIsValidType(valueType, valueTextValue, validTypes, parserConfig, customTypes) {
  return validTypes.includes("Any") || validTypes.includes(valueType) || validTypes.includes("String") && valueType === "Identifier" || validTypes.includes("Number") && valueType === "BigInt" || isValidIPAddressType(valueType, valueTextValue, validTypes, parserConfig) || isValidUIDType(valueType, valueTextValue, validTypes, parserConfig) || isValidSmartscapeIdType(
    valueType,
    valueTextValue,
    validTypes,
    parserConfig
  ) || isValidTimestampType(valueType, valueTextValue, validTypes, parserConfig) || !!isValidCustomType(valueTextValue, customTypes);
}
function hasValuePredicate(entry) {
  return !!entry && typeof entry === "object" && typeof entry.valuePredicate === "function";
}
function isValidCustomType(textValue, customTypes) {
  if (!customTypes) {
    return false;
  }
  for (const [typeName, validatorFn] of Object.entries(customTypes)) {
    if (typeof validatorFn === "function") {
      if (validatorFn(textValue)) {
        return typeName;
      }
      continue;
    }
    if (hasValuePredicate(validatorFn) && validatorFn.valuePredicate(textValue)) {
      return typeName;
    }
  }
  return false;
}
function isValidOperator(textValue, key, internalValidatorMap, parserConfig) {
  const operatorName = (0, import_operator_specification.getOperatorName)(textValue);
  if (operatorName === "") {
    return false;
  }
  if (operatorName === "search" && !parserConfig?.searchConversion) {
    return true;
  }
  return getValidOperators(key, internalValidatorMap, parserConfig).includes(
    operatorName
  );
}
function isValidValue(key, valueTextValue, valueType, internalValidatorMap, currentStatement, filterFieldValue, customTypes, parserConfig) {
  if (!key || !internalValidatorMap) {
    return { isValid: true };
  }
  const predicate = internalValidatorMap.normalizedKeys.byKey.get(key);
  if (!predicate) {
    return { isValid: true };
  }
  const validValues = predicate.valuePredicate;
  if (typeof validValues === "function") {
    const result = validValues({
      currentStatement,
      filterFieldValue
    });
    if (typeof result === "boolean") {
      return { isValid: result };
    }
    return { isValid: false, message: result };
  }
  if (valueType === "Variable") {
    return { isValid: true };
  }
  let validTypes;
  let hasLiteralValues;
  if (predicate.valueType) {
    validTypes = Array.isArray(predicate.valueType) ? predicate.valueType : [predicate.valueType];
    hasLiteralValues = Array.isArray(validValues);
  } else if (validValues !== void 0) {
    ({ types: validTypes, hasLiteralValues } = (0, import_suggestion_type.getValidValueTypes)(validValues));
  } else {
    return { isValid: true };
  }
  const isValidType = checkIsValidType(
    valueType,
    valueTextValue,
    validTypes,
    parserConfig,
    customTypes
  );
  if (Array.isArray(validValues)) {
    let textMatches = false;
    for (const validValue of validValues) {
      if ((0, import_type_guards.isFilterFieldValidatorMapTypeSpecification)(validValue)) {
        if (!isValidType && !hasLiteralValues || isValidType) {
          return { isValid: isValidType, validTypes };
        }
      } else if ((0, import_type_guards.isFilterFieldValueSuggestionGroupConfig)(validValue)) {
        if (validValue.suggestions.some((suggestion) => {
          if ((0, import_type_guards.isFilterFieldValidatorMapTypeSpecification)(suggestion)) {
            return false;
          }
          return literalMatchesTextValue(valueTextValue, suggestion);
        })) {
          textMatches = true;
        }
        continue;
      } else if (literalMatchesTextValue(valueTextValue, validValue)) {
        textMatches = true;
      }
    }
    return {
      isValid: textMatches || isWildCardOperator(currentStatement),
      validTypes
    };
  }
  if ((0, import_type_guards.isFilterFieldValidatorMapTypeSpecification)(validValues)) {
    return { isValid: isValidType, validTypes };
  }
  if ((0, import_type_guards.isFilterFieldValueSuggestionGroupConfig)(validValues)) {
    const groupSuggestionMatches = validValues.suggestions.some(
      (suggestion) => {
        if ((0, import_type_guards.isFilterFieldValidatorMapTypeSpecification)(suggestion)) {
          return false;
        }
        return literalMatchesTextValue(valueTextValue, suggestion);
      }
    );
    return {
      isValid: groupSuggestionMatches || isWildCardOperator(currentStatement)
    };
  }
  if (validValues === void 0) {
    return { isValid: isValidType };
  }
  return {
    isValid: literalMatchesTextValue(valueTextValue, validValues),
    validTypes
  };
}
function literalMatchesTextValue(textValue, predicate) {
  if ((0, import_type_guards.isFilterFieldValidatorMapDurationSpecification)(predicate)) {
    return textValue === `${predicate.value}${predicate.unit}`;
  }
  if ((0, import_type_guards.isFilterFieldRichValuePredicate)(predicate)) {
    const cleaned2 = `${predicate.value}`.replaceAll(/(?<!\\)(?:\\\\)*(\\)(?!\\)/g, "").replaceAll(/(\\\\)/g, "\\");
    return textValue === cleaned2;
  }
  const cleaned = `${predicate}`.replaceAll(/(?<!\\)(?:\\\\)*(\\)(?!\\)/g, "").replaceAll(/(\\\\)/g, "\\");
  return textValue === cleaned;
}
function getValuePredicateTextValue(predicate) {
  if ((0, import_type_guards.isFilterFieldValidatorMapDurationSpecification)(predicate)) {
    return `${predicate.value}${predicate.unit}`;
  } else if ((0, import_type_guards.isFilterFieldRichValuePredicate)(predicate)) {
    return `${predicate.value}`;
  } else {
    return `${predicate}`;
  }
}
function getValidOperators(key, internalValidatorMap, parserConfig) {
  key = typeof key === "string" ? key.replace(/^"(.*?(?<!\\)(?:\\\\)*)"$/, "$1") : key;
  const defaultOperators = [
    ...import_operator_specification.FILTERFIELD_DEFAULT_OPERATORS,
    // eslint-disable-next-line @typescript-eslint/no-deprecated
    ...parserConfig?.searchConversion ? ["search"] : []
  ];
  if (!internalValidatorMap) {
    return defaultOperators;
  }
  const globalOperators = internalValidatorMap.comparisonOperators || defaultOperators;
  const predicate = key ? internalValidatorMap.normalizedKeys.byKey.get(key) : void 0;
  if (!predicate && key === "*") {
    return [];
  }
  if (!predicate) {
    return globalOperators;
  }
  if (predicate.comparisonOperators) {
    return predicate.comparisonOperators;
  }
  let allowedTypes;
  if (predicate.valueType) {
    allowedTypes = Array.isArray(predicate.valueType) ? predicate.valueType : [predicate.valueType];
  } else if (!predicate.valuePredicate || typeof predicate.valuePredicate === "function") {
    allowedTypes = "Any";
  } else {
    allowedTypes = (0, import_suggestion_type.getValidValueTypes)(predicate.valuePredicate).types;
  }
  if (allowedTypes !== "Any") {
    return intersectOperators(
      globalOperators,
      allowedTypes.reduce((acc, type) => {
        acc.push(...(0, import_operator_specification.getAllowedOperatorsForType)(type));
        return acc;
      }, [])
    );
  }
  return globalOperators;
}
const OPERATOR_SHORT_NAME_MAP = {
  equals: "=",
  "not-equals": "!=",
  "less-than": "<",
  "less-or-equal": "<=",
  "greater-than": ">",
  "greater-or-equal": ">=",
  "matches-phrase": "~",
  "not-matches-phrase": "!~",
  in: "in",
  "not in": "not in",
  contains: "contains",
  "starts-with": "starts-with",
  "ends-with": "ends-with",
  "not-starts-with": "not-starts-with",
  "not-contains": "not-contains",
  "not-ends-with": "not-ends-with",
  exists: "= *",
  "not-exists": "!= *"
};
function getValidOperatorsShort(key, internalValidatorMap, parserConfig) {
  const operators = getValidOperators(key, internalValidatorMap, parserConfig);
  const result = [];
  for (const [key2, value] of Object.entries(OPERATOR_SHORT_NAME_MAP)) {
    if (operators.includes(key2)) {
      result.push(value);
    }
  }
  return result;
}
function getValidOperatorsColored(key, internalValidatorMap, parserConfig) {
  return getValidOperatorsShort(key, internalValidatorMap, parserConfig).map(
    (operator) => `<span style="color: ${import_colors.default.Syntax.Operators}">${operator}</span>`
  ).join(", ");
}
function getColoredSearchOperator() {
  return `<span style="color: ${import_colors.default.Syntax.Variable}">*</span> <span style="color: ${import_colors.default.Syntax.Operators}">~</span>`;
}
function intersectOperators(globalOperators, operators) {
  return globalOperators.filter((op) => operators.includes(op));
}
