// @dynatrace/strato-components/filters/filter-field/utils/auto-format-logical-operators.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var auto_format_logical_operators_exports = {};
__export(auto_format_logical_operators_exports, {
  autoFormatLogicalOperators: () => autoFormatLogicalOperators
});
module.exports = __toCommonJS(auto_format_logical_operators_exports);
var import_state = require("@codemirror/state");
function autoFormatLogicalOperators(value, tree, cursorPosition) {
  let newValue = value;
  const transactions = [];
  const logicalOperators = value.matchAll(/\b(and|or)\b/gi);
  for (const operator of logicalOperators) {
    const { index = -1 } = operator;
    if (index === -1) {
      continue;
    }
    if (tree.cursorAt(index, 1).name === "LogicalOperator") {
      const endIndex = index + operator[0].length;
      const capitalize = value.substring(endIndex, endIndex + 1) === " " || endIndex + 1 === cursorPosition && tree.cursorAt(endIndex + 1, 1).name === "LogicalOperator";
      const _operator = value.substring(index, endIndex);
      const newDoc = capitalize ? _operator.toUpperCase() : _operator;
      transactions.push({
        changes: import_state.ChangeSet.of(
          {
            from: index,
            to: endIndex,
            insert: newDoc
          },
          value.length
        ),
        selection: { anchor: cursorPosition },
        sequential: true
      });
      newValue = `${newValue.substring(0, index)}${capitalize ? _operator.toUpperCase() : _operator}${newValue.substring(endIndex)}`;
    }
  }
  return [newValue, transactions];
}
