// @dynatrace/strato-components/filters/filter-field/utils/generate-errors-or-diagnostics.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var generate_errors_or_diagnostics_exports = {};
__export(generate_errors_or_diagnostics_exports, {
  generateErrorsOrDiagnostics: () => generateErrorsOrDiagnostics
});
module.exports = __toCommonJS(generate_errors_or_diagnostics_exports);
var import_state = require("@codemirror/state");
var import_decoration_marks = require("../decorations/decoration-marks.js");
var import_editor_config = require("../editor-state-configs/editor-config.js");
const getRenderMessage = (message) => {
  const messageWrapperElement = document.createElement("div");
  messageWrapperElement.className = "strato-filterfield-error-wrapper";
  messageWrapperElement.innerHTML = `<div class="strato-filterfield-error-icon-wrapper"><svg focusable="false"
  role="img"
  aria-hidden="true"
  fill="currentColor"
  width="16"
  height="16"
  xmlns="http://www.w3.org/2000/svg"
  viewBox="0 0 20 20"><path d="M9.2799 6.26743H10.7799V10.7674H9.2799V6.26743Z" />
  <path d="M11.0299 12.7674C11.0299 12.2151 10.5822 11.7674 10.0299 11.7674C9.47762 11.7674 9.0299 12.2151 9.0299 12.7674C9.0299 13.3197 9.47762 13.7674 10.0299 13.7674C10.5822 13.7674 11.0299 13.3197 11.0299 12.7674Z" />
  <path d="M10.737 1.54289C10.3465 1.15237 9.71332 1.15237 9.32279 1.54289L1.54289 9.32279C1.15237 9.71332 1.15237 10.3465 1.54289 10.737L9.32279 18.5169C9.71332 18.9074 10.3465 18.9074 10.737 18.5169L18.5169 10.737C18.9074 10.3465 18.9074 9.71332 18.5169 9.32279L10.737 1.54289ZM10.0299 2.95711L17.1027 10.0299L10.0299 17.1027L2.95711 10.0299L10.0299 2.95711Z" /></svg></div><div>${message}</div>`;
  return () => messageWrapperElement;
};
function generateErrorsOrDiagnostics(view, messages) {
  const builder = new import_state.RangeSetBuilder();
  const diagnostics = [];
  const errorsStateField = view.state.field(import_editor_config.usedErrors, false);
  const pastingStateField = view?.state.field(import_editor_config.pastingMode, false);
  if (!errorsStateField || view.visibleRanges.length === 0) {
    if (messages !== void 0) {
      builder.finish();
      return diagnostics;
    }
    return builder.finish();
  }
  for (const error of errorsStateField.errors) {
    const {
      range: { from, to },
      decoration,
      message: errorMessage
    } = error;
    let message;
    let messageDetails;
    if (typeof errorMessage === "string") {
      message = errorMessage;
    } else {
      message = errorMessage.type;
      messageDetails = errorMessage.details;
    }
    if (from <= view.visibleRanges[0].to && to <= view.visibleRanges[0].to) {
      const showError = !isErrorInPastingRange(from, to, pastingStateField);
      if (showError) {
        builder.add(from, to, decoration ?? import_decoration_marks.errorMark);
      }
      if (messages !== void 0) {
        let errorMessage2 = message;
        const errorType = messages[message];
        if (errorType) {
          if (typeof errorType === "function") {
            errorMessage2 = errorType(
              messageDetails
            );
          } else {
            errorMessage2 = errorType;
          }
        }
        if (showError) {
          diagnostics.push({
            from,
            to,
            severity: "error",
            message: "",
            renderMessage: getRenderMessage(errorMessage2)
          });
        }
      }
    }
  }
  if (messages !== void 0) {
    builder.finish();
    return diagnostics;
  }
  return builder.finish();
}
function isErrorInPastingRange(errorFrom, errorTo, pastingStateField) {
  if (!pastingStateField?.range || !pastingStateField.pastingInProgress) {
    return false;
  }
  const pastingRangeFrom = pastingStateField.range.from;
  const pastingRangeTo = pastingStateField.range.to;
  return (
    // Check if error starts inside the pasting range
    // Check if error surrounds the pasting range
    errorFrom >= pastingRangeFrom && errorFrom <= pastingRangeTo || // Check if error starts inside the pasting range
    errorTo >= pastingRangeFrom && errorTo <= pastingRangeFrom || errorFrom <= pastingRangeFrom && errorTo >= pastingRangeTo
  );
}
