// @dynatrace/strato-components/filters/filter-field/utils/get-suggestion-transaction.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var get_suggestion_transaction_exports = {};
__export(get_suggestion_transaction_exports, {
  _isFollowedByOpeningBracket: () => _isFollowedByOpeningBracket,
  getSuggestionTransaction: () => getSuggestionTransaction,
  insertionIsLastListValue: () => insertionIsLastListValue
});
module.exports = __toCommonJS(get_suggestion_transaction_exports);
var import_language = require("@codemirror/language");
var import_state = require("@codemirror/state");
var import_escape_suggestion = require("./escape-suggestion.js");
var import_get_statement_status = require("./get-statement-status.js");
var import_get_text = require("./get-text.js");
var import_tree_navigation = require("./tree-navigation.js");
var import_tree_node_guards = require("./tree-node-guards.js");
var import_filterfield_language = require("../parser/filterfield-language.js");
function getSuggestionTransaction(suggestionValue, range, shouldEscapeSuggestion, trailing, insertionStrategy, cursorBeforeTrailing, suggestionsContext, editorState) {
  const { token, currentStatement: _currentStatement } = suggestionsContext;
  const currentSelection = editorState.selection;
  const { head } = currentSelection.main;
  let currentStatement = _currentStatement;
  if (!currentStatement) {
    currentStatement = {
      type: "Statement",
      range: { from: token.range.from, to: token.range.to }
    };
  }
  const autoEscapeSuggestion = insertionStrategy === "replace-token" && shouldEscapeSuggestion === void 0 || shouldEscapeSuggestion;
  const currentDocumentLength = editorState.doc.toString().length;
  const isWildcardSuggestionWithValuePlaceholder = trailing?.includes("value");
  const isInclusionOperatorSuggestion = ["in (", "not in ("].includes(
    suggestionValue
  );
  const isSuggestionWithWildcard = ["= *", "!= *"].includes(suggestionValue);
  const nextStatement = (0, import_tree_navigation.hasNextStatement)(head, editorState);
  if (isInclusionOperatorSuggestion && range === void 0) {
    return getSuggestionTransactionForInclusionOperatorSuggestion(
      suggestionValue,
      currentStatement,
      token,
      trailing,
      head,
      currentDocumentLength,
      editorState,
      nextStatement
    );
  }
  const appliedSuggestionValue = `${autoEscapeSuggestion && !isWildcardSuggestionWithValuePlaceholder ? (0, import_escape_suggestion.escapeSuggestion)(suggestionValue) : suggestionValue}${trailing ?? ""}`;
  const trimmedValue = appliedSuggestionValue.trimEnd();
  const isExistsOperator = trimmedValue.match(/^(= \*)$|^(!= \*)$/);
  const addSpace = appliedSuggestionValue.slice(-1) === " ";
  let insertedValue = appliedSuggestionValue;
  let replaceFrom = token.range.from;
  let replaceUntil = token.range.to;
  let nextSibling = false;
  let isLastListValue = false;
  let startsWithOperator = false;
  let endsWithOperator = false;
  let hasSpaceAfterQuotes = false;
  const followedByOpeningParenthesis = _isFollowedByOpeningBracket(
    editorState,
    head
  );
  if (token.type === "List" && head > token.range.from && head < token.range.to && insertionStrategy === "replace-token") {
    replaceFrom = head;
    replaceUntil = head;
  }
  const { value } = currentStatement ?? {};
  const isValueToken = token.range.from === value?.range.from && token.range.to === value?.range.to;
  if (currentStatement) {
    nextSibling = !nextStatement ? (0, import_tree_navigation.hasNextSibling)(head, editorState, { ignorePositionOperators: true }) : nextStatement;
    startsWithOperator = isValueToken && currentStatement.comparisonOperator?.type === "ComparisonOperator" && (currentStatement.comparisonOperator.value === "starts-with" || currentStatement.comparisonOperator.value === "contains");
    endsWithOperator = isValueToken && currentStatement.comparisonOperator?.type === "ComparisonOperator" && (currentStatement.comparisonOperator.value === "ends-with" || currentStatement.comparisonOperator.value === "contains");
    if (nextSibling && insertionIsLastListValue(head, editorState)) {
      isLastListValue = true;
    }
    if (insertionStrategy === "replace-statement" || isExistsOperator && currentStatement.value && !followedByOpeningParenthesis || startsWithOperator) {
      nextSibling = nextStatement;
      replaceUntil = currentStatement.range.to;
    }
    insertedValue = nextSibling || insertionStrategy === "insert" ? trimmedValue : appliedSuggestionValue;
    const containsOperator = startsWithOperator && endsWithOperator;
    if ((startsWithOperator || endsWithOperator) && insertionStrategy === "replace-token") {
      let { from, to } = {
        from: currentStatement.range.to,
        to: currentStatement.range.to
      };
      if (currentStatement.value) {
        ({ from, to } = currentStatement.value.range);
      }
      if (startsWithOperator && (head > from && head < to || !containsOperator && head === from)) {
        startsWithOperator = false;
        if (!trimmedValue.endsWith("*")) {
          insertedValue = containsOperator && !trimmedValue.startsWith("*") ? `*${trimmedValue}*` : `${trimmedValue}*`;
          insertedValue = `${insertedValue}${!nextSibling && addSpace ? " " : ""}`;
        }
      } else {
        if (head === from || startsWithOperator && head === to) {
          replaceFrom = from;
          insertedValue = appliedSuggestionValue;
        }
        replaceUntil = to;
        startsWithOperator = false;
        nextSibling = nextStatement;
      }
    }
  } else {
    insertedValue = nextSibling ? trimmedValue : appliedSuggestionValue;
  }
  let insertionRange;
  switch (insertionStrategy) {
    case "insert":
      insertionRange = {
        from: head,
        to: head
      };
      break;
    case "replace-statement": {
      const editorState2 = import_state.EditorState.create({
        doc: appliedSuggestionValue,
        extensions: [import_filterfield_language.filterFieldLanguageExtension]
      });
      const suggestionCursor = (0, import_tree_navigation.getStatementCursorAtPosition)(
        appliedSuggestionValue.length,
        editorState2
      );
      if ((0, import_get_statement_status.getStatementStatus)(suggestionCursor) !== "complete") {
        insertedValue = appliedSuggestionValue;
        nextSibling = false;
      }
      insertionRange = {
        from: Math.min(
          currentDocumentLength,
          currentStatement?.range.from ?? head
        ),
        to: Math.min(currentDocumentLength, currentStatement?.range.to ?? head)
      };
      if ((0, import_tree_node_guards.isLogicalOperator)(token)) {
        nextSibling = nextStatement;
        if (nextSibling) {
          insertedValue = trimmedValue;
        }
        insertionRange = {
          from: Math.min(currentDocumentLength, token.range.from),
          to: Math.min(currentDocumentLength, token.range.to)
        };
      }
      break;
    }
    case "replace-all":
      insertedValue = appliedSuggestionValue;
      nextSibling = false;
      insertionRange = {
        from: 0,
        to: editorState.doc.toString().length
      };
      break;
    default: {
      if ((token.type === "String" || token.type === "Key" || token.type === "Error") && range === void 0) {
        if (token.range.to < (currentStatement?.range.to ?? 0)) {
          const characterAfterToken = (0, import_get_text.getTextFromRange)(
            {
              from: token.range.to,
              to: token.range.to + 1
            },
            editorState.doc.toString()
          );
          if (characterAfterToken === " ") {
            nextSibling = true;
            insertedValue = trimmedValue;
          }
        }
        if (head > token.range.from && head < token.range.to) {
          const isTokenEscaped = token.textValue.match(/^"[^"]*"$/);
          const isAppliedSuggestionValueEscaped = appliedSuggestionValue.match(/^"[^"]*"$/) ?? appliedSuggestionValue.match(/^"[^"]*" $/);
          if (isTokenEscaped && !isAppliedSuggestionValueEscaped) {
            replaceFrom = token.range.from + 1;
            replaceUntil = token.range.to;
            const addSpace2 = insertedValue.slice(-1) !== " ";
            insertedValue = addSpace2 ? `${trimmedValue}" ` : `${trimmedValue}"`;
            hasSpaceAfterQuotes = addSpace2;
          }
        }
      } else if (range) {
        ({ from: replaceFrom, to: replaceUntil } = range);
      }
      insertionRange = {
        from: Math.min(currentDocumentLength, replaceFrom),
        to: Math.min(currentDocumentLength, replaceUntil)
      };
    }
  }
  insertionRange = getSortedRange(insertionRange);
  let anchor = insertionStrategy === "insert" ? head + insertedValue.length : Math.min(currentDocumentLength, insertionRange.from) + (nextSibling && !isLastListValue ? appliedSuggestionValue.length : insertedValue.length) + (insertionStrategy === "replace-token" ? +startsWithOperator : 0) + +hasSpaceAfterQuotes;
  if (cursorBeforeTrailing || isWildcardSuggestionWithValuePlaceholder) {
    const trailingLength = trailing ? trailing.length : 0;
    anchor -= trailingLength + (autoEscapeSuggestion && !(isSuggestionWithWildcard || isWildcardSuggestionWithValuePlaceholder) ? 1 : 0);
  }
  const newDocLength = currentDocumentLength - (insertionRange.to - insertionRange.from) + insertedValue.length;
  anchor = Math.max(0, Math.min(anchor, newDocLength));
  return {
    changes: {
      ...insertionRange,
      insert: insertedValue
    },
    selection: {
      anchor,
      // Select the value placeholder if inserting a wildcard suggestion with a placeholder
      ...isWildcardSuggestionWithValuePlaceholder ? { head: anchor + 5 } : {}
      // 5 = 'value'.length;
    }
  };
}
function getSuggestionTransactionForInclusionOperatorSuggestion(suggestionValue, currentStatement, token, trailing, head, currentDocumentLength, editorState, nextStatement) {
  let insertedValue = suggestionValue;
  if (currentStatement?.value) {
    let stringifiedValue = currentStatement.value.textValue;
    stringifiedValue = stringifiedValue.replace(/^\(/, "").replace(/(?<!\\)\)$/, "");
    if (currentStatement.value.type !== "List") {
      insertedValue = `${insertedValue}${stringifiedValue.trimEnd()}`;
    } else {
      insertedValue = `${insertedValue}${stringifiedValue}`;
    }
  }
  let insertionRange = {
    from: Math.min(
      currentDocumentLength,
      currentStatement?.comparisonOperator?.range.from ?? token.range.from
    ),
    to: Math.min(
      currentDocumentLength,
      currentStatement?.value?.range.to ?? token.range.to
    )
  };
  insertionRange = getSortedRange(insertionRange);
  let anchor = insertionRange.from + insertedValue.length;
  insertedValue = `${insertedValue}${trailing}`;
  const nextSibling = !nextStatement ? (0, import_tree_navigation.hasNextSibling)(head, editorState, { ignorePositionOperators: true }) : nextStatement;
  insertedValue = `${insertedValue.trimEnd()}${!nextSibling ? " " : ""}`;
  const newDocLength = currentDocumentLength - (insertionRange.to - insertionRange.from) + insertedValue.length;
  anchor = Math.max(0, Math.min(anchor, newDocLength));
  return {
    changes: {
      ...insertionRange,
      insert: insertedValue
    },
    selection: {
      anchor
    }
  };
}
function getSortedRange(range) {
  if (range.to < range.from) {
    return { from: range.to, to: range.from };
  }
  return range;
}
function _isFollowedByOpeningBracket(editorState, head) {
  const tree = (0, import_language.syntaxTree)(editorState);
  let cursor = tree.cursorAt(head, 1);
  if (cursor.name === "OpeningBracket") {
    return true;
  }
  if (cursor.node.name === "Filter" || cursor.node.name === "FilterStatement") {
    cursor = tree.cursorAt(head, -1);
  }
  if (!cursor.next()) {
    return false;
  }
  if (cursor.name === "OpeningBracket" && cursor.node.from >= head) {
    return true;
  }
  if (cursor.name === "\u26A0") {
    if (cursor.next() && cursor.node.from >= head) {
      if (cursor.name === "OpeningBracket" || cursor.name === "FilterGroup") {
        return true;
      }
    }
  }
  if (cursor.name === "FilterIncludesValue") {
    if (cursor.next() && cursor.node.from >= head) {
      if (cursor.name === "OpeningBracket" || cursor.name === "List") {
        return true;
      }
    }
  }
  const cursorAtExactPosition = tree.cursorAt(head + 1, 1);
  if (cursorAtExactPosition.node.name === "OpeningBracket") {
    return true;
  }
  cursorAtExactPosition.parent();
  if (cursorAtExactPosition.node.name !== "Filter" && cursor.node.name === "FilterGroup") {
    return true;
  }
  return false;
}
function insertionIsLastListValue(cursorPosition, editorState) {
  const tree = (0, import_language.syntaxTree)(editorState);
  const cursor = tree.cursorAt(cursorPosition, 1);
  if (cursor.node.name === "ClosingBracket" && cursorPosition === cursor.node.from) {
    if (cursor.parent() && cursor.name === "List") {
      return true;
    }
  }
  return false;
}
