// @dynatrace/strato-components/filters/filter-field/utils/migrate-validator-map.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var migrate_validator_map_exports = {};
__export(migrate_validator_map_exports, {
  migrateFilterFieldKeyPredicatesToArray: () => migrateFilterFieldKeyPredicatesToArray
});
module.exports = __toCommonJS(migrate_validator_map_exports);
var import_type_guards = require("./validation-utils/type-guards.js");
function migrateValuePredicate(valuePredicate) {
  if (typeof valuePredicate === "function") {
    return { valuePredicate };
  }
  if ((0, import_type_guards.isFilterFieldValidatorMapTypeSpecification)(valuePredicate)) {
    const raw = valuePredicate.type;
    const types = Array.isArray(raw) ? raw : [raw];
    const valueType = normalizeCollectedTypes([...new Set(types)]);
    return { valueType };
  }
  if ((0, import_type_guards.isFilterFieldValidatorMapDurationSpecification)(valuePredicate)) {
    return { valuePredicate: `${valuePredicate.value}${valuePredicate.unit}` };
  }
  if ((0, import_type_guards.isFilterFieldValueSuggestionGroupConfig)(valuePredicate)) {
    return {
      valuePredicate: {
        ...valuePredicate,
        suggestions: valuePredicate.suggestions.map(migrateValuePredicateItem)
      }
    };
  }
  if (Array.isArray(valuePredicate)) {
    return migrateValuePredicateArray(valuePredicate);
  }
  return { valuePredicate };
}
function migrateFilterFieldKeyPredicatesToArray(validatorMap) {
  const { keyPredicates } = validatorMap;
  if (keyPredicates === void 0 || Array.isArray(keyPredicates)) {
    return validatorMap;
  }
  const result = Object.entries(
    keyPredicates
  ).map(([key, config]) => {
    if ((0, import_type_guards.isFilterFieldValidatorMapTypeSpecification)(config)) {
      const raw = config.type;
      const types = Array.isArray(raw) ? raw : [raw];
      return { key, valueType: normalizeCollectedTypes([...new Set(types)]) };
    }
    const { valuePredicate, ...rest } = config;
    if (valuePredicate === void 0) {
      return { key, ...rest };
    }
    const { valueType, valuePredicate: migratedPredicate } = migrateValuePredicate(valuePredicate);
    return {
      key,
      ...rest,
      ...valueType !== void 0 && { valueType },
      ...migratedPredicate !== void 0 && {
        valuePredicate: migratedPredicate
      }
    };
  });
  return { ...validatorMap, keyPredicates: result };
}
function migrateValuePredicateItem(item) {
  if ((0, import_type_guards.isFilterFieldValidatorMapDurationSpecification)(item)) {
    return `${item.value}${item.unit}`;
  }
  return item;
}
function normalizeCollectedTypes(types) {
  if (types.length === 0) {
    return void 0;
  }
  return types.length === 1 ? types[0] : types;
}
function migrateValuePredicateArray(items) {
  const collectedTypes = [];
  const remaining = [];
  for (const item of items) {
    if ((0, import_type_guards.isFilterFieldValidatorMapTypeSpecification)(item)) {
      const types = Array.isArray(item.type) ? item.type : [item.type];
      for (const t of types) {
        if (!collectedTypes.includes(t)) {
          collectedTypes.push(t);
        }
      }
    } else if ((0, import_type_guards.isFilterFieldValueSuggestionGroupConfig)(item)) {
      remaining.push({
        ...item,
        suggestions: item.suggestions.map(migrateValuePredicateItem)
      });
    } else {
      remaining.push(migrateValuePredicateItem(item));
    }
  }
  const valueType = normalizeCollectedTypes(collectedTypes);
  const valuePredicate = remaining.length === 0 ? void 0 : remaining.length === 1 ? remaining[0] : remaining;
  return { valueType, valuePredicate };
}
