// @dynatrace/strato-components/filters/filter-field/utils/tree-node-guards.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var tree_node_guards_exports = {};
__export(tree_node_guards_exports, {
  isFilterFieldBigIntNode: () => isFilterFieldBigIntNode,
  isFilterFieldBooleanNode: () => isFilterFieldBooleanNode,
  isFilterFieldComparisonOperatorNode: () => isFilterFieldComparisonOperatorNode,
  isFilterFieldDurationNode: () => isFilterFieldDurationNode,
  isFilterFieldErrorNode: () => isFilterFieldErrorNode,
  isFilterFieldExistsOperatorNode: () => isFilterFieldExistsOperatorNode,
  isFilterFieldGroupNode: () => isFilterFieldGroupNode,
  isFilterFieldInclusionOperatorNode: () => isFilterFieldInclusionOperatorNode,
  isFilterFieldJSONPathNode: () => isFilterFieldJSONPathNode,
  isFilterFieldKeyNode: () => isFilterFieldKeyNode,
  isFilterFieldLeafNode: () => isFilterFieldLeafNode,
  isFilterFieldListNode: () => isFilterFieldListNode,
  isFilterFieldNumberNode: () => isFilterFieldNumberNode,
  isFilterFieldSearchOperatorNode: () => isFilterFieldSearchOperatorNode,
  isFilterFieldStatementNode: () => isFilterFieldStatementNode,
  isFilterFieldStringNode: () => isFilterFieldStringNode,
  isFilterFieldSuggestion: () => isFilterFieldSuggestion,
  isFilterFieldSuggestionGroup: () => isFilterFieldSuggestionGroup,
  isFilterFieldValueNode: () => isFilterFieldValueNode,
  isFilterFieldVariableNode: () => isFilterFieldVariableNode,
  isFilterValue: () => isFilterValue,
  isLogicalOperator: () => isLogicalOperator,
  isValidPositionOperator: () => isValidPositionOperator
});
module.exports = __toCommonJS(tree_node_guards_exports);
function isFilterFieldJSONPathNode(node) {
  return typeof node === "object" && node.type === "JSONPath";
}
function isFilterFieldStringNode(node) {
  return typeof node === "object" && node.type === "String";
}
function isFilterFieldVariableNode(node) {
  return typeof node === "object" && node.type === "Variable";
}
function isFilterFieldNumberNode(node) {
  return typeof node === "object" && node.type === "Number";
}
function isFilterFieldBigIntNode(node) {
  return typeof node === "object" && node.type === "BigInt";
}
function isFilterFieldBooleanNode(node) {
  return typeof node === "object" && node.type === "Boolean";
}
function isFilterFieldDurationNode(node) {
  return typeof node === "object" && node.type === "Duration";
}
function isFilterFieldExistsOperatorNode(node) {
  return typeof node === "object" && node.type === "ExistsOperator";
}
function isFilterFieldSearchOperatorNode(node) {
  return typeof node === "object" && node.type === "SearchOperator";
}
function isFilterFieldInclusionOperatorNode(node) {
  return typeof node === "object" && node.type === "InclusionOperator";
}
function isFilterFieldErrorNode(node) {
  return typeof node === "object" && node.type === "Error";
}
function isFilterFieldKeyNode(node) {
  return typeof node === "object" && node.type === "Key";
}
function isFilterFieldComparisonOperatorNode(node) {
  return typeof node === "object" && node.type === "ComparisonOperator";
}
function isFilterFieldGroupNode(node) {
  return typeof node === "object" && node.children !== void 0;
}
function isFilterFieldStatementNode(node) {
  return typeof node === "object" && node.type === "Statement";
}
function isFilterFieldListNode(node) {
  return typeof node === "object" && node.textValue !== void 0 && Array.isArray(node.value);
}
function isFilterFieldValueNode(node) {
  return typeof node === "object" && node.textValue !== void 0;
}
function isFilterFieldLeafNode(node) {
  return typeof node === "object" && node.textValue !== void 0 && !Array.isArray(node.value);
}
function isFilterValue(token) {
  const valueTypes = [
    "String",
    "Number",
    "Boolean",
    "Duration",
    "List",
    "Error"
  ];
  return valueTypes.includes(token.type);
}
function isValidPositionOperator(node) {
  return node.comparisonOperator?.value === "=" || node.comparisonOperator?.value === "!=" || node.comparisonOperator?.value === "ends-with" || node.comparisonOperator?.value === "not-ends-with";
}
function isLogicalOperator(node) {
  return node && (node.type === "LogicalOperator" || node.type === "Error" && (node.value === "OR" || node.value === "AND"));
}
function isFilterFieldSuggestionGroup(value) {
  return typeof value === "object" && (value.label !== void 0 || value.suggestions !== void 0);
}
function isFilterFieldSuggestion(value) {
  const suggestionValue = value;
  return typeof value === "object" && (suggestionValue.value !== void 0 || suggestionValue.displayValue !== void 0 || suggestionValue.insertionStrategy !== void 0);
}
