// @dynatrace/strato-components/charts/xy-chart-adoption/timeseries/utils/render-series.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var render_series_exports = {};
__export(render_series_exports, {
  defineAxisId: () => defineAxisId,
  renderAreaSeries: () => renderAreaSeries,
  renderBandSeries: () => renderBandSeries,
  renderBarSeries: () => renderBarSeries,
  renderLineSeries: () => renderLineSeries,
  xAxisId: () => xAxisId
});
module.exports = __toCommonJS(render_series_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_get_unit_as_string = require("../../../core/utils/get-unit-as-string.js");
var import_XYChartRoot = require("../../../xy-chart/XYChartRoot.js");
const xAxisId = "x-axis";
const defineAxisId = (unit) => `y-axis-${(0, import_get_unit_as_string.getUnitAsString)(unit)}`;
const baseSeriesProps = ({ data, unit, seriesId }) => ({
  data,
  xAxisId,
  yAxisId: defineAxisId(unit),
  x0Accessor: "start",
  x1Accessor: "end",
  seriesIdAccessor: "seriesId",
  nameAccessor: "seriesName",
  key: seriesId
});
const renderLineSeries = (series) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartRoot.XYChart.LineSeries, { ...baseSeriesProps(series), y0Accessor: "value" });
const renderAreaSeries = (series) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartRoot.XYChart.AreaSeries, { ...baseSeriesProps(series), y1Accessor: "value" });
const renderBarSeries = (series) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
  import_XYChartRoot.XYChart.BarSeries,
  {
    ...baseSeriesProps(series),
    variant: "stack",
    y1Accessor: "value"
  }
);
const renderBandSeries = (series) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
  import_XYChartRoot.XYChart.AreaSeries,
  {
    ...baseSeriesProps(series),
    y0Accessor: "y0",
    y1Accessor: "y1"
  }
);
