// @dynatrace/strato-components/charts/core/types/chart-types.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var chart_types_exports = {};
__export(chart_types_exports, {
  ChartType: () => ChartType
});
module.exports = __toCommonJS(chart_types_exports);
var ChartType = /* @__PURE__ */ ((ChartType2) => {
  ChartType2["AnnotationsChart"] = "annotations-chart";
  ChartType2["CategoricalBar"] = "categorical-bar";
  ChartType2["Pie"] = "pie";
  ChartType2["Donut"] = "donut";
  ChartType2["SingleValue"] = "single-value";
  ChartType2["TimeSeries"] = "time-series";
  ChartType2["Histogram"] = "histogram";
  ChartType2["SingleMeterBar"] = "single-meter-bar";
  ChartType2["MultiMeterBar"] = "multi-meter-bar";
  ChartType2["Honeycomb"] = "honeycomb";
  ChartType2["Treemap"] = "treemap";
  ChartType2["XYChart"] = "xy-chart";
  ChartType2["GaugeChart"] = "gauge-chart";
  return ChartType2;
})(ChartType || {});
