// @dynatrace/strato-components/charts/core/types/color-palette.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var color_palette_exports = {};
__export(color_palette_exports, {
  apdexMap: () => apdexMap,
  automaticColoringOptions: () => automaticColoringOptions,
  colorPaletteModeOptions: () => colorPaletteModeOptions,
  colorPaletteOptions: () => colorPaletteOptions,
  divergingColoringOptions: () => divergingColoringOptions,
  isCustomColorPalette: () => isCustomColorPalette,
  isCustomColorRecordPalette: () => isCustomColorRecordPalette,
  logLevelMap: () => logLevelMap,
  logLevelWithAlias: () => logLevelWithAlias,
  logStatusMap: () => logStatusMap,
  logStatusWithAlias: () => logStatusWithAlias,
  securityRiskLevelMap: () => securityRiskLevelMap,
  sequentialColoringOptions: () => sequentialColoringOptions,
  shouldApplyAutomatically: () => shouldApplyAutomatically,
  statusMap: () => statusMap,
  themedColoringOptions: () => themedColoringOptions,
  thresholdMap: () => thresholdMap,
  vulnerabilityStatusMap: () => vulnerabilityStatusMap
});
module.exports = __toCommonJS(color_palette_exports);
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
const themedColoringOptions = [
  "categorical",
  "purple-rain",
  "swamps",
  "fireplace",
  "blue-steel",
  "rainbow",
  "categorical-inverted",
  "purple-rain-inverted",
  "swamps-inverted",
  "fireplace-inverted",
  "blue-steel-inverted",
  "rainbow-inverted"
];
const automaticColoringOptions = [
  "log-status",
  "log-level",
  "apdex",
  "security-risk-level",
  "vulnerability-status",
  "threshold",
  "status",
  "log-status-inverted",
  "log-level-inverted",
  "apdex-inverted",
  "security-risk-level-inverted",
  "vulnerability-status-inverted",
  "threshold-inverted",
  "status-inverted"
];
const sequentialColoringOptions = [
  "pink-purple",
  "blue-turquoise",
  "blue",
  "grey",
  "moss",
  "turquoise",
  "pink",
  "yellow",
  "petrol",
  "purple",
  "pear",
  "brown",
  "red",
  "magenta",
  "orange",
  "pink-purple-inverted",
  "blue-turquoise-inverted",
  "blue-inverted",
  "grey-inverted",
  "moss-inverted",
  "turquoise-inverted",
  "pink-inverted",
  "yellow-inverted",
  "petrol-inverted",
  "purple-inverted",
  "pear-inverted",
  "brown-inverted",
  "red-inverted",
  "magenta-inverted",
  "orange-inverted"
];
const divergingColoringOptions = [
  "orange-blue",
  "red-blue",
  "red-purple",
  "blue-moss",
  "purple-yellow",
  "red-green",
  "orange-petrol",
  "pink-green",
  "orange-blue-inverted",
  "red-blue-inverted",
  "red-purple-inverted",
  "blue-moss-inverted",
  "purple-yellow-inverted",
  "red-green-inverted",
  "orange-petrol-inverted",
  "pink-green-inverted"
];
const colorPaletteOptions = [
  ...themedColoringOptions,
  ...automaticColoringOptions,
  ...sequentialColoringOptions,
  ...divergingColoringOptions
];
const colorPaletteModeOptions = ["single-color", "multi-color"];
function isCustomColorPalette(colorPalette) {
  return typeof colorPalette === "object" && colorPalette !== null && (!Array.isArray(colorPalette) || colorPalette.length > 0);
}
function isCustomColorRecordPalette(colorPalette) {
  return !Array.isArray(colorPalette);
}
const shouldApplyAutomatically = (colorPalette) => {
  return automaticColoringOptions.includes(colorPalette);
};
const logStatusMap = {
  error: import_colors.default.Charts.Logstatus.Error.Default,
  warning: import_colors.default.Charts.Logstatus.Warning.Default,
  info: import_colors.default.Charts.Logstatus.Info.Default,
  none: import_colors.default.Charts.Logstatus.None.Default,
  warn: import_colors.default.Charts.Logstatus.Warning.Default
  // alias for warning
};
const logLevelMap = {
  emergency: import_colors.default.Charts.Loglevel.Emergency.Default,
  error: import_colors.default.Charts.Loglevel.Error.Default,
  alert: import_colors.default.Charts.Loglevel.Alert.Default,
  critical: import_colors.default.Charts.Loglevel.Critical.Default,
  severe: import_colors.default.Charts.Loglevel.Severe.Default,
  warning: import_colors.default.Charts.Loglevel.Warning.Default,
  notice: import_colors.default.Charts.Loglevel.Notice.Default,
  info: import_colors.default.Charts.Loglevel.Info.Default,
  debug: import_colors.default.Charts.Loglevel.Debug.Default,
  none: import_colors.default.Charts.Loglevel.None.Default,
  warn: import_colors.default.Charts.Loglevel.Warning.Default
  // alias for warning
};
const logStatusWithAlias = {
  error: import_colors.default.Charts.Logstatus.Error,
  warning: import_colors.default.Charts.Logstatus.Warning,
  info: import_colors.default.Charts.Logstatus.Info,
  none: import_colors.default.Charts.Logstatus.None,
  warn: import_colors.default.Charts.Logstatus.Warning
  // alias for warning
};
const logLevelWithAlias = {
  emergency: import_colors.default.Charts.Loglevel.Emergency,
  error: import_colors.default.Charts.Loglevel.Error,
  alert: import_colors.default.Charts.Loglevel.Alert,
  critical: import_colors.default.Charts.Loglevel.Critical,
  severe: import_colors.default.Charts.Loglevel.Severe,
  warning: import_colors.default.Charts.Loglevel.Warning,
  notice: import_colors.default.Charts.Loglevel.Notice,
  info: import_colors.default.Charts.Loglevel.Info,
  debug: import_colors.default.Charts.Loglevel.Debug,
  none: import_colors.default.Charts.Loglevel.None,
  warn: import_colors.default.Charts.Loglevel.Warning
  // alias for warning
};
const apdexMap = {
  excellent: import_colors.default.Charts.Apdex.Excellent.Default,
  good: import_colors.default.Charts.Apdex.Good.Default,
  fair: import_colors.default.Charts.Apdex.Fair.Default,
  poor: import_colors.default.Charts.Apdex.Poor.Default,
  unacceptable: import_colors.default.Charts.Apdex.Unacceptable.Default
};
const securityRiskLevelMap = {
  muted: import_colors.default.Charts.SecurityRiskLevel.Muted.Default,
  critical: import_colors.default.Charts.SecurityRiskLevel.Critical.Default,
  high: import_colors.default.Charts.SecurityRiskLevel.High.Default,
  medium: import_colors.default.Charts.SecurityRiskLevel.Medium.Default,
  low: import_colors.default.Charts.SecurityRiskLevel.Low.Default,
  none: import_colors.default.Charts.SecurityRiskLevel.None.Default
};
const vulnerabilityStatusMap = {
  muted: import_colors.default.Charts.VulnerabilityStatus.Muted.Default,
  resolved: import_colors.default.Charts.VulnerabilityStatus.Resolved.Default,
  affected: import_colors.default.Charts.VulnerabilityStatus.Affected.Default
};
const thresholdMap = {
  good: import_colors.default.Charts.Threshold.Good.Default,
  warning: import_colors.default.Charts.Threshold.Warning.Default,
  bad: import_colors.default.Charts.Threshold.Bad.Default
};
const statusMap = {
  ideal: import_colors.default.Charts.Status.Ideal.Default,
  good: import_colors.default.Charts.Status.Good.Default,
  neutral: import_colors.default.Charts.Status.Neutral.Default,
  warning: import_colors.default.Charts.Status.Warning.Default,
  critical: import_colors.default.Charts.Status.Critical.Default
};
