// @dynatrace/strato-components/charts/core/types/color-palettes-definitions.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var color_palettes_definitions_exports = {};
__export(color_palettes_definitions_exports, {
  automaticColoringMap: () => automaticColoringMap,
  colorPaletteMap: () => colorPaletteMap,
  colorPalettesWithAlias: () => colorPalettesWithAlias,
  divergingColoringMap: () => divergingColoringMap,
  sequentialColoringMap: () => sequentialColoringMap
});
module.exports = __toCommonJS(color_palettes_definitions_exports);
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
var import_color_palette = require("./color-palette.js");
var import_prepare_inverted_color_palette = require("../utils/prepare-inverted-color-palette.js");
const colorPalettesWithAlias = (0, import_prepare_inverted_color_palette.prepareInvertedColorPalette)({
  "log-status": import_color_palette.logStatusWithAlias,
  "log-level": import_color_palette.logLevelWithAlias
});
const themedColoringMap = (0, import_prepare_inverted_color_palette.prepareInvertedColorPalette)({
  categorical: import_colors.default.Charts.Categorical,
  "purple-rain": import_colors.default.Charts.CategoricalThemed.PurpleRain,
  swamps: import_colors.default.Charts.CategoricalThemed.Swamps,
  fireplace: import_colors.default.Charts.CategoricalThemed.Fireplace,
  "blue-steel": import_colors.default.Charts.CategoricalThemed.BlueSteel,
  rainbow: import_colors.default.Charts.Rainbow
});
const internalAutomaticColoringMap = (0, import_prepare_inverted_color_palette.prepareInvertedColorPalette)({
  apdex: import_colors.default.Charts.Apdex,
  "security-risk-level": import_colors.default.Charts.SecurityRiskLevel,
  "vulnerability-status": import_colors.default.Charts.VulnerabilityStatus,
  threshold: import_colors.default.Charts.Threshold,
  status: import_colors.default.Charts.Status
});
const automaticColoringMap = (0, import_prepare_inverted_color_palette.prepareInvertedColorPalette)({
  "log-status": import_color_palette.logStatusMap,
  "log-level": import_color_palette.logLevelMap,
  apdex: import_color_palette.apdexMap,
  "security-risk-level": import_color_palette.securityRiskLevelMap,
  "vulnerability-status": import_color_palette.vulnerabilityStatusMap,
  threshold: import_color_palette.thresholdMap,
  status: import_color_palette.statusMap
});
const sequentialColoringMap = (0, import_prepare_inverted_color_palette.prepareInvertedColorPalette)({
  "pink-purple": import_colors.default.Charts.Sequential.PinkPurple,
  "blue-turquoise": import_colors.default.Charts.Sequential.BlueTurquoise,
  blue: import_colors.default.Charts.Sequential.Blue,
  grey: import_colors.default.Charts.Sequential.Grey,
  moss: import_colors.default.Charts.Sequential.Moss,
  turquoise: import_colors.default.Charts.Sequential.Turquoise,
  pink: import_colors.default.Charts.Sequential.Pink,
  yellow: import_colors.default.Charts.Sequential.Yellow,
  petrol: import_colors.default.Charts.Sequential.Petrol,
  purple: import_colors.default.Charts.Sequential.Purple,
  pear: import_colors.default.Charts.Sequential.Pear,
  brown: import_colors.default.Charts.Sequential.Brown,
  red: import_colors.default.Charts.Sequential.Red,
  magenta: import_colors.default.Charts.Sequential.Magenta,
  orange: import_colors.default.Charts.Sequential.Orange
});
const divergingColoringMap = (0, import_prepare_inverted_color_palette.prepareInvertedColorPalette)({
  "orange-blue": import_colors.default.Charts.Diverging.OrangeBlue,
  "red-blue": import_colors.default.Charts.Diverging.RedBlue,
  "red-purple": import_colors.default.Charts.Diverging.RedPurple,
  "blue-moss": import_colors.default.Charts.Diverging.BlueMoss,
  "purple-yellow": import_colors.default.Charts.Diverging.PurpleYellow,
  "red-green": import_colors.default.Charts.Diverging.RedGreen,
  "orange-petrol": import_colors.default.Charts.Diverging.OrangePetrol,
  "pink-green": import_colors.default.Charts.Diverging.PinkGreen
});
const colorPaletteMap = {
  ...themedColoringMap,
  ...internalAutomaticColoringMap,
  ...colorPalettesWithAlias,
  ...sequentialColoringMap,
  ...divergingColoringMap
};
