// @dynatrace/strato-components/charts/core/types/pointer-events.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var pointer_events_exports = {};
__export(pointer_events_exports, {
  PointerEventActions: () => PointerEventActions,
  isDragMove: () => isDragMove,
  isMiddleButtonDragMove: () => isMiddleButtonDragMove,
  isPinchGesture: () => isPinchGesture,
  isPinchStartGesture: () => isPinchStartGesture,
  isSwipeMove: () => isSwipeMove,
  isSwipeStart: () => isSwipeStart
});
module.exports = __toCommonJS(pointer_events_exports);
var PointerEventActions = /* @__PURE__ */ ((PointerEventActions2) => {
  PointerEventActions2["Idle"] = "idle";
  PointerEventActions2["Hover"] = "hover";
  PointerEventActions2["Click"] = "click";
  PointerEventActions2["DblClick"] = "dbl-click";
  PointerEventActions2["Tap"] = "tap";
  PointerEventActions2["DblTap"] = "dbl-tap";
  PointerEventActions2["DragStart"] = "drag:start";
  PointerEventActions2["DragMove"] = "drag:move";
  PointerEventActions2["DragEnd"] = "drag:end";
  PointerEventActions2["PinchStart"] = "pinch:start";
  PointerEventActions2["Pinch"] = "pinch";
  PointerEventActions2["PinchEnd"] = "pinch:end";
  PointerEventActions2["SwipeStart"] = "swipe:start";
  PointerEventActions2["SwipeMove"] = "swipe:move";
  PointerEventActions2["SwipeEnd"] = "swipe:end";
  PointerEventActions2["MiddleButtonDragStart"] = "middle-button-drag:start";
  PointerEventActions2["MiddleButtonDragMove"] = "middle-button-drag:move";
  PointerEventActions2["MiddleButtonDragEnd"] = "middle-button-drag:end";
  return PointerEventActions2;
})(PointerEventActions || {});
const isDragMove = (gesture) => gesture?.action === "drag:move" /* DragMove */;
const isMiddleButtonDragMove = (gesture) => gesture?.action === "middle-button-drag:move" /* MiddleButtonDragMove */;
const isPinchGesture = (gesture) => gesture?.action === "pinch" /* Pinch */;
const isPinchStartGesture = (gesture) => gesture?.action === "pinch:start" /* PinchStart */;
const isSwipeStart = (gesture) => gesture?.action === "swipe:start" /* SwipeStart */;
const isSwipeMove = (gesture) => gesture?.action === "swipe:move" /* SwipeMove */;
