// @dynatrace/strato-components/charts/core/types/type-guards.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var type_guards_exports = {};
__export(type_guards_exports, {
  isBoundedRange: () => isBoundedRange,
  isDynamic: () => isDynamic,
  isDynamicNumeric: () => isDynamicNumeric,
  isDynamicNumericRange: () => isDynamicNumericRange,
  isDynamicRange: () => isDynamicRange,
  isPointThreshold: () => isPointThreshold,
  isStatic: () => isStatic,
  isStaticRange: () => isStaticRange
});
module.exports = __toCommonJS(type_guards_exports);
function isStatic(data) {
  return data !== null && !Array.isArray(data) && "value" in data;
}
function isStaticRange(data) {
  return data !== null && !Array.isArray(data) && "min" in data && "max" in data;
}
function isDynamic(data) {
  return data !== null && Array.isArray(data) && data[0] && "value" in data[0] && "date" in data[0];
}
function isDynamicNumeric(data) {
  return data !== null && Array.isArray(data) && data[0] && "value" in data[0] && "point" in data[0];
}
function isDynamicRange(data) {
  return data !== null && Array.isArray(data) && data[0] && "min" in data[0] && "max" in data[0] && "date" in data[0];
}
function isDynamicNumericRange(data) {
  return data !== null && Array.isArray(data) && data[0] && "min" in data[0] && "max" in data[0] && "point" in data[0];
}
function isPointThreshold(y0, y1) {
  return y0.valueOf() === y1.valueOf();
}
function isBoundedRange(data) {
  return data !== null && !Array.isArray(data) && "value" in data;
}
