// @dynatrace/strato-components/charts/core/types/zoom.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var zoom_exports = {};
__export(zoom_exports, {
  ZOOM_ACTIONS: () => ZOOM_ACTIONS,
  ZOOM_CHANGE_TYPE: () => ZOOM_CHANGE_TYPE
});
module.exports = __toCommonJS(zoom_exports);
const ZOOM_CHANGE_TYPE = Object.freeze({
  Reset: "reset",
  /**
   * @deprecated ZoomX is deprecated, please use `<[Chart].Zoom/>`  instead.
   */
  ZoomX: "zoom-x",
  Panning: "panning",
  ZoomIn: "zoom-in",
  ZoomToFit: "zoom-to-fit",
  ZoomOut: "zoom-out"
});
const ZOOM_ACTIONS = Object.freeze({
  ZoomIn: "zoom-in",
  ZoomOut: "zoom-out",
  Reset: "reset"
});
