// @dynatrace/strato-components/content/expandable-text/ExpandableText.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ExpandableText_exports = {};
__export(ExpandableText_exports, {
  ExpandableText: () => ExpandableText
});
module.exports = __toCommonJS(ExpandableText_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react2 = require("react");
var import_react_intl = require("react-intl");
var import_ExpandableText_css = require("./ExpandableText.sty.js");
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useControllableState = require("../../core/hooks/useControllableState.js");
var import_useId = require("../../core/hooks/useId.js");
var import_useSafeForwardProps = require("../../core/hooks/useSafeForwardProps.js");
const COMPONENT_NAME = "ExpandableText";
const ExpandableText = (0, import_react2.forwardRef)(
  (props, forwardedRef) => {
    const {
      children,
      expandLabel = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_react_intl.FormattedMessage,
        {
          description: "Shown for expandable text.",
          defaultMessage: "Show more",
          id: "DV3d5lBJda+EHJja"
        }
      ),
      collapseLabel = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_react_intl.FormattedMessage,
        {
          description: "Shown for collapsible text.",
          defaultMessage: "Show less",
          id: "986buSbL34GVFlBf"
        }
      ),
      defaultExpanded = false,
      expanded,
      onExpandChange,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
    const ariaLabelingProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(restProps);
    const forwardProps = (0, import_useSafeForwardProps.useSafeForwardProps)(restProps, ariaLabelingProps);
    const [isExpanded, setExpanded] = (0, import_useControllableState.useControllableState)({
      defaultValue: defaultExpanded,
      value: expanded,
      onChange: onExpandChange
    });
    const textId = (0, import_useId.useId)();
    const showMoreLessDataDtProperties = `state:${isExpanded ? "less" : "more"}`;
    const beginsWithWhitespace = Array.isArray(children) === false && typeof children === "string" && /^\s/.test(children);
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
        "span",
        {
          ref: forwardedRef,
          id: textId,
          role: "region",
          className: (0, import_clsx.default)(
            (0, import_ExpandableText_css.expandableTextCSS)({ expanded: isExpanded }),
            consumerClassName
          ),
          style: consumerStyle,
          "data-testid": dataTestId,
          "data-dtrum-mask": dataDtrumMask,
          "data-dtrum-allow": dataDtrumAllow,
          ...ariaLabelingProps,
          ...behavioralTrackingProps,
          ...forwardProps,
          "data-dt-component": COMPONENT_NAME,
          children: [
            isExpanded && beginsWithWhitespace && " ",
            isExpanded && children
          ]
        }
      ),
      " ",
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        "button",
        {
          onClick: () => setExpanded(!isExpanded),
          "aria-controls": textId,
          "aria-expanded": isExpanded,
          className: import_ExpandableText_css.expandableTextButtonCSS,
          "data-dtrum-mask": dataDtrumMask,
          "data-dtrum-allow": dataDtrumAllow,
          "data-dt-name": "ShowMoreLess",
          "data-dt-properties": showMoreLessDataDtProperties,
          children: isExpanded ? collapseLabel : expandLabel
        }
      )
    ] });
  }
);
ExpandableText.displayName = COMPONENT_NAME;
