// @dynatrace/strato-components/charts/histogram/components/overlays/HistogramSelectionArea/HistogramSelectionArea.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var HistogramSelectionArea_exports = {};
__export(HistogramSelectionArea_exports, {
  HistogramSelectionArea: () => HistogramSelectionArea
});
module.exports = __toCommonJS(HistogramSelectionArea_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_HistogramSelectionAreaTooltip = require("./HistogramSelectionAreaTooltip.js");
var import_HistogramSelectionDatapoints = require("./HistogramSelectionDatapoints.js");
var import_useGridPlotSize = require("../../../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_explore_reducers = require("../../../../core/components/interactions-state-helpers/reducers/explore-reducers.js");
var import_DynamicSelectionArea = require("../../../../core/components/selection-area/DynamicSelectionArea.js");
var import_get_percentage_from_pixel = require("../../../../core/utils/get-percentage-from-pixel.js");
var import_get_selection_pixels = require("../../../../core/utils/get-selection-pixels.js");
var import_selectors = require("../../../hooks/store/selectors.js");
var import_state_setters = require("../../../hooks/store/state-setters.js");
var import_useHistogramChartOverlayRef = require("../../../hooks/useHistogramChartOverlayRef.js");
var import_useHistogramInteractions = require("../../../hooks/useHistogramInteractions.js");
var import_useHistogramSelectionAreaRef = require("../../../hooks/useHistogramSelectionAreaRef.js");
var import_useHistogramXFormatter = require("../../../hooks/useHistogramXFormatter.js");
var import_useHistogramXScale = require("../../../hooks/useHistogramXScale.js");
var import_useXAxisConfig = require("../../../hooks/useXAxisConfig.js");
const HistogramSelectionArea = () => {
  const bottomFormatter = (0, import_useHistogramXFormatter.useHistogramXFormatter)();
  const selectionAreaRef = (0, import_useHistogramSelectionAreaRef.useHistogramSelectionAreaRef)();
  const { unit: xUnit } = (0, import_useXAxisConfig.useXAxisConfig)();
  const { xScale } = (0, import_useHistogramXScale.useHistogramXScale)();
  const chartRef = (0, import_useHistogramChartOverlayRef.useHistogramChartOverlayRef)();
  const setState = (0, import_state_setters.useSetState)();
  const selectState = (0, import_selectors.useSelectState)();
  const { width } = (0, import_useGridPlotSize.useGridPlotSize)();
  const { zoomX } = (0, import_useHistogramInteractions.useHistogramInteractions)();
  const hasMagnifiers = !(0, import_lodash_es.isUndefined)(bottomFormatter);
  if ((0, import_lodash_es.isUndefined)(selectState) || !chartRef.current) {
    return null;
  }
  const { start, end, selectionPercentage, finished, accessible, variant } = selectState;
  const handleWrapperKeyDown = (event) => {
    const canZoom = zoomX?.enabled;
    if (event.key === "Enter" && canZoom) {
      chartRef.current?.focus();
    }
  };
  const handleSelectionChange = ({
    selection,
    accessible: accessible2
  }) => {
    setState(
      (0, import_explore_reducers.reduceExploreSelect)({
        start: selection[0],
        end: selection[1],
        finished: true,
        accessible: accessible2,
        variant,
        selectionPercentage: (0, import_get_percentage_from_pixel.getPercentageFromPixel)(selection, width)
      })
    );
  };
  const selectionPosition = finished ? (0, import_get_selection_pixels.getSelectionPixels)(selectionPercentage, width, [start, end]) : [start, end];
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    variant === "select" ? /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_HistogramSelectionDatapoints.HistogramSelectionDatapoints, {}),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_HistogramSelectionAreaTooltip.HistogramSelectionAreaTooltip, {})
    ] }) : null,
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_DynamicSelectionArea.DynamicSelectionArea,
      {
        chartRef,
        selection: selectionPosition,
        hasMagnifiers,
        variant,
        focusable: accessible,
        interactable: finished,
        bottomFormatter: (value) => bottomFormatter(xScale.invert(value), xUnit),
        onKeyDown: handleWrapperKeyDown,
        onSelectionChange: handleSelectionChange,
        selectionAreaRef
      }
    )
  ] });
};
