// @dynatrace/strato-components/charts/histogram/components/overlays/HistogramSelectionArea/HistogramSelectionAreaTooltip.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var HistogramSelectionAreaTooltip_exports = {};
__export(HistogramSelectionAreaTooltip_exports, {
  HistogramSelectionAreaTooltip: () => HistogramSelectionAreaTooltip
});
module.exports = __toCommonJS(HistogramSelectionAreaTooltip_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_useGridPlotSize = require("../../../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_get_custom_context_menu_items = require("../../../../core/components/interactions-overlay/utils/get-custom-context-menu-items.js");
var import_OverlayTooltip = require("../../../../core/components/overlay-tooltip/OverlayTooltip.js");
var import_SelectionTooltipWithAsyncActions = require("../../../../core/components/selection-area/SelectionTooltipWithAsyncActions.js");
var import_get_selection_tooltip_position = require("../../../../core/components/selection-area/tooltip/get-selection-tooltip-position.js");
var import_usePointerAndScrollResponsiveChartBounds = require("../../../../core/hooks/usePointerAndScrollResponsiveChartBounds.js");
var import_selectors = require("../../../hooks/store/selectors.js");
var import_useHistogramChartOverlayRef = require("../../../hooks/useHistogramChartOverlayRef.js");
var import_useHistogramFilteredData = require("../../../hooks/useHistogramFilteredData.js");
var import_useHistogramInteractions = require("../../../hooks/useHistogramInteractions.js");
var import_useHistogramSelectConfig = require("../../../hooks/useHistogramSelectConfig.js");
var import_useHistogramXFormatter = require("../../../hooks/useHistogramXFormatter.js");
var import_useHistogramXScale = require("../../../hooks/useHistogramXScale.js");
var import_useZoomToSelection = require("../../../hooks/useZoomToSelection.js");
var import_tooltip_text = require("../../../messages/tooltip-text.js");
var import_is_bin_within_range = require("../utils/is-bin-within-range.js");
const HistogramSelectionAreaTooltip = () => {
  const selectState = (0, import_selectors.useSelectState)();
  const zoomToSelection = (0, import_useZoomToSelection.useZoomToSelection)();
  const { zoomX } = (0, import_useHistogramInteractions.useHistogramInteractions)();
  const filteredData = (0, import_useHistogramFilteredData.useHistogramFilteredData)();
  const { xScale } = (0, import_useHistogramXScale.useHistogramXScale)();
  const formatter = (0, import_useHistogramXFormatter.useHistogramXFormatter)();
  const selectConfig = (0, import_useHistogramSelectConfig.useHistogramSelectConfig)();
  const { width } = (0, import_useGridPlotSize.useGridPlotSize)();
  const intl = (0, import_react_intl.useIntl)();
  const chartRef = (0, import_useHistogramChartOverlayRef.useHistogramChartOverlayRef)();
  const chartBounds = (0, import_usePointerAndScrollResponsiveChartBounds.usePointerAndScrollResponsiveChartBounds)(
    chartRef,
    true
  )();
  if (!selectState || !chartBounds) {
    return null;
  }
  const selectionStart = selectState.selectionPercentage ? width * selectState.selectionPercentage[0] / 100 : selectState.start.relative.x;
  const selectionEnd = selectState.selectionPercentage ? width * selectState.selectionPercentage[1] / 100 : selectState.end.relative.x;
  const range = [
    xScale.invert(selectionStart),
    xScale.invert(selectionEnd)
  ];
  const selectedSeries = [];
  let totalBins = 0;
  const [from, to] = range;
  for (const series of filteredData.series) {
    let isSeriesPresent = false;
    const seriesBins = [];
    for (const bin of series.bins) {
      if ((0, import_is_bin_within_range.isBinWithinRange)(bin, from, to)) {
        totalBins++;
        seriesBins.push(bin);
        isSeriesPresent = true;
      }
    }
    if (isSeriesPresent) {
      selectedSeries.push({ ...series, bins: seriesBins });
    }
  }
  const tooltipPosition = (0, import_get_selection_tooltip_position.getSelectionTooltipPosition)(selectState, chartBounds);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_SelectionTooltipWithAsyncActions.SelectionTooltipWithAsyncActions,
    {
      actions: () => {
        const actions = selectConfig?.actions;
        return actions && (async () => actions(selectedSeries, range))().then(
          import_get_custom_context_menu_items.getCustomContextMenuItems
        );
      },
      selectFinished: selectState.finished,
      header: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_OverlayTooltip.OverlayTooltip.Header,
        {
          header1: [formatter(from), formatter(to)],
          duration1: formatter(Math.abs(to - from)),
          subheader: `${totalBins} bins across ${selectedSeries.length} series`
        }
      ),
      position: tooltipPosition,
      zoomEnabled: !!zoomX?.enabled,
      range,
      children: zoomX?.enabled ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_OverlayTooltip.OverlayTooltip.Action,
        {
          label: intl.formatMessage(import_tooltip_text.HISTOGRAM_MESSAGES.zoomToSelection),
          icon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ZoomToSelectionIcon, {}),
          onSelect: zoomToSelection,
          trailingIcon: "\u23CE"
        }
      ) : null
    }
  );
};
