// @dynatrace/strato-components/esm/charts/xy-chart/components/tooltip/ClosestPointTooltip/closestPointTooltipBodyUtils.js
import { safeGet } from "../../../../core/utils/safe-get.js";
import {
  isAreaTooltipContext
} from "../../../utils/type-guards.js";
import { toRelative } from "../../../utils/value-representation.js";
import { getTooltipFormatter } from "../utils/get-tooltip-formatter.js";
import { getClosestPointTooltipBodyContent } from "./hooks/getClosestPointTooltipBodyContent.js";
const buildClosestPointTooltipBodyProps = ({
  datapoint,
  series,
  seriesToMetadata,
  axisIdToFormatter,
  axisIdToScaleDomain,
  datapointMetadataMap,
  intl
}) => {
  const seriesMetadata = safeGet(seriesToMetadata, series);
  const enrichedDatapoint = {
    series,
    datapoint,
    metadata: seriesMetadata
  };
  const { xAxis, yAxis } = enrichedDatapoint.metadata;
  const xFormatter = getTooltipFormatter(xAxis, axisIdToFormatter, intl);
  const yFormatter = getTooltipFormatter(yAxis, axisIdToFormatter, intl);
  const datapointMetadata = safeGet(datapointMetadataMap, datapoint);
  const binDomain = datapointMetadata?.binDomain;
  let xDomain = safeGet(axisIdToScaleDomain, xAxis.id);
  let yDomain = safeGet(axisIdToScaleDomain, yAxis.id);
  if (isAreaTooltipContext(enrichedDatapoint)) {
    const { isStacked } = enrichedDatapoint.metadata;
    if (isStacked && datapointMetadata.baseline !== "x" && binDomain) {
      xDomain = { min: binDomain * -1, max: binDomain };
    }
    if (isStacked && datapointMetadata.baseline !== "y" && binDomain) {
      yDomain = { min: binDomain * -1, max: binDomain };
    }
  }
  const { text: textX, duration: durationX } = getClosestPointTooltipBodyContent(
    [datapoint.x0, datapoint.x1],
    xFormatter,
    xAxis.isRelative ? (value) => toRelative(value, xDomain) : void 0
  );
  const { text: textY, duration: durationY } = getClosestPointTooltipBodyContent(
    [datapoint.y0, datapoint.y1],
    yFormatter,
    yAxis.isRelative ? (value) => toRelative(value, yDomain) : void 0
  );
  return {
    xAxisLabel: xAxis.label,
    yAxisLabel: yAxis.label,
    xAxisValue: textX,
    yAxisValue: textY,
    xAxisDuration: durationX,
    yAxisDuration: durationY
  };
};
export {
  buildClosestPointTooltipBodyProps
};
