// @dynatrace/strato-components/esm/charts/xy-chart/components/tooltip/ClosestPointTooltip/closestPointTooltipHeaderUtils.js
import { getDefaultAction } from "../../../../core/components/series-action/hooks/useDefaultAction.js";
import { safeGet } from "../../../../core/utils/safe-get.js";
import { buildAreaSeriesActionItem } from "../../../utils/build-series-action-item.js";
import {
  isAreaTooltipContext
} from "../../../utils/type-guards.js";
const buildClosestPointTooltipHeaderProps = ({
  datapoint,
  series,
  seriesToMetadata,
  seriesWithDatapoints,
  datapointMetadataMap,
  truncationMode,
  idFactory = () => crypto.randomUUID()
}) => {
  const metadata = safeGet(seriesToMetadata, series);
  const enrichedDatapoint = {
    series,
    datapoint,
    metadata
  };
  const { name = "", color = { Default: "blue" } } = metadata;
  const shape = "shape" in metadata ? metadata.shape : void 0;
  const content = {
    id: idFactory(),
    name
  };
  let actions = void 0;
  if (isAreaTooltipContext(enrichedDatapoint)) {
    const { actions: seriesActions } = enrichedDatapoint.metadata;
    actions = seriesActions ? () => {
      const actionItem = buildAreaSeriesActionItem({
        datapoint: enrichedDatapoint.datapoint,
        allSeries: seriesWithDatapoints,
        seriesToMetadata,
        datapointMetadataMap,
        currentSeries: enrichedDatapoint.series
      });
      return seriesActions(actionItem);
    } : void 0;
  } else {
    const { actions: seriesActions } = enrichedDatapoint.metadata;
    actions = () => seriesActions?.({
      ...datapoint,
      name
    });
  }
  return {
    color: color.Default,
    name,
    actions,
    defaultActions: getDefaultAction(content),
    truncationMode,
    shape
  };
};
export {
  buildClosestPointTooltipHeaderProps
};
