// @dynatrace/strato-components/esm/charts/xy-chart/components/tooltip/ClosestPointTooltip/ClosestPointTooltip.js
import { jsx, jsxs } from "react/jsx-runtime";
import { useIntl } from "react-intl";
import { buildClosestPointTooltipBodyProps } from "./closestPointTooltipBodyUtils.js";
import { buildClosestPointTooltipHeaderProps } from "./closestPointTooltipHeaderUtils.js";
import { OverlayTooltip } from "../../../../core/components/overlay-tooltip/OverlayTooltip.js";
import { useAreaData } from "../../../hooks/useAreaData.js";
import { useAxesFormatter } from "../../../hooks/useAxesFormatter.js";
import { useXYChartData } from "../../../hooks/useXYChartData.js";
import { useXYChartTruncationMode } from "../../../hooks/useXYChartTruncationMode.js";
function ClosestPointTooltip({
  datapoint,
  series,
  seriesToMetadata
}) {
  const axisIdToFormatter = useAxesFormatter();
  const { axisIdToScaleDomain, datapointMetadataMap } = useXYChartData();
  const intl = useIntl();
  const { filteredSeriesWithDatapoints } = useAreaData();
  const { truncationMode } = useXYChartTruncationMode();
  const { name, color, actions, shape } = buildClosestPointTooltipHeaderProps({
    datapoint,
    series,
    seriesToMetadata,
    seriesWithDatapoints: filteredSeriesWithDatapoints,
    datapointMetadataMap
  });
  const { xAxisValue, yAxisValue, xAxisDuration, yAxisDuration } = buildClosestPointTooltipBodyProps({
    datapoint,
    series,
    seriesToMetadata,
    axisIdToFormatter,
    axisIdToScaleDomain,
    datapointMetadataMap,
    intl
  });
  const data = [
    {
      name,
      value: `${yAxisValue} ${yAxisDuration ? `(${yAxisDuration})` : ""}`,
      color,
      shape
    }
  ];
  const label = datapoint.x0 instanceof Date ? "" : void 0;
  return /* @__PURE__ */ jsxs(OverlayTooltip, { truncation: truncationMode, children: [
    /* @__PURE__ */ jsx(
      OverlayTooltip.Header,
      {
        header1: xAxisValue,
        duration1: xAxisDuration,
        label
      }
    ),
    /* @__PURE__ */ jsx(
      OverlayTooltip.Series,
      {
        data,
        variant: "shared",
        activeItem: -1,
        seriesItemTemplate: (item) => /* @__PURE__ */ jsx(OverlayTooltip.SeriesItem, { ...item, actions })
      }
    )
  ] });
}
export {
  ClosestPointTooltip
};
