// @dynatrace/strato-geo/esm/map/components/toolbar/constants.js
import { defineMessages } from "react-intl";
const initialA11yToolbarTabIndexContext = [
  { "zoom-in": 0 },
  { "zoom-out": -1 },
  { reset: -1 },
  { "zoom-to-fit": -1 },
  { "download-data": -1 }
];
const TOOLBAR_MESSAGES = defineMessages({
  zoomActionZoomIn: {
    id: "11eDVd4MkXWk8Bok",
    defaultMessage: "Zoom in",
    description: 'Text in a tooltip shown when hovering the "Zoom in" action in the geo map chart toolbar'
  },
  zoomActionZoomOut: {
    id: "5DTYqeFoXLMaXUHR",
    defaultMessage: "Zoom out",
    description: 'Text in a tooltip shown when hovering the "Zoom out" action in the geo map chart toolbar'
  },
  zoomActionReset: {
    id: "hVE7x0G0Q7/ih93A",
    defaultMessage: "Reset",
    description: 'Text in a tooltip shown when hovering the "Reset" action in the geo map chart toolbar'
  },
  zoomActionZoomToFit: {
    id: "EhcEZc1+Xc9wYsXg",
    defaultMessage: "Zoom to fit",
    description: 'Text in a tooltip shown when hovering the "Zoom to fit" action in the geo map chart toolbar'
  },
  downloadData: {
    id: "hlwWx47trlDMOy+V",
    defaultMessage: "Download Data as CSV",
    description: 'Text in a tooltip shown when hovering the "Download Data" action in the geo map chart toolbar'
  }
});
export {
  TOOLBAR_MESSAGES,
  initialA11yToolbarTabIndexContext
};
