// @dynatrace/strato-geo/esm/map/components/toolbar/MapToolbar.js
import { jsx, jsxs } from "react/jsx-runtime";
import { useMap } from "@vis.gl/react-maplibre";
import {
  _ChartToolbarRenderer as ChartToolbarRenderer,
  _useOverlayTooltipReducer as useTooltipReducer
} from "@dynatrace/strato-components/charts";
import { MapDownloadDataButton } from "./buttons/MapDownloadDataButton.js";
import { MapResetButton } from "./buttons/MapResetButton.js";
import { MapZoomInOutButtons } from "./buttons/MapZoomInOutButtons.js";
import { MapZoomToFitButton } from "./buttons/MapZoomToFitButton.js";
import { initialA11yToolbarTabIndexContext } from "./constants.js";
import { useMapConfig } from "../../hooks/use-map-config.js";
import { useMapRawData } from "../../hooks/use-map-raw-data.js";
import { MapInitialViewProvider } from "../../providers/map-initial-view.provider.js";
import { useMapOverlayState } from "../../store/selectors.js";
import { useSetState } from "../../store/store.js";
import { buildAndDownloadCsv } from "../../utils/build-and-download-csv.js";
const defaultToolbarPlacement = "top-right";
const MapToolbar = () => {
  const { toolbar: toolbarConfig, interactions: interactionsConfig } = useMapConfig();
  const data = useMapRawData();
  const { current: map } = useMap();
  const { mouseInBounds } = useMapOverlayState();
  const setState = useSetState();
  const dispatch = useTooltipReducer();
  const isZoomEnabled = toolbarConfig?.zoom?.enabled ?? interactionsConfig?.zoom?.enabled ?? false;
  const isZoomToFitEnabled = toolbarConfig?.zoom?.enabled ?? interactionsConfig?.zoomToFit?.enabled ?? false;
  const isDownloadDataEnabled = toolbarConfig?.downloadCSV?.enabled ?? false;
  const hasZoomControls = isZoomEnabled || isZoomToFitEnabled;
  const hasControls = hasZoomControls || isDownloadDataEnabled;
  const hasOnlyDownloadCSV = isDownloadDataEnabled && !hasZoomControls;
  const handleMouseEnter = () => {
    dispatch({ type: "RESET_TOOLTIP" })();
    if (map) {
      map.queryRenderedFeatures().forEach((feature) => {
        if (feature.id !== void 0 && feature.layer.source) {
          map.setFeatureState(
            { source: feature.layer.source, id: feature.id },
            { active: false, isAnyActive: false }
          );
        }
      });
    }
    setState((prev) => ({
      ...prev
    }));
  };
  const handleMouseLeave = () => {
    setState((prev) => ({
      ...prev,
      tooltip: { ...prev.tooltip, enabled: true }
    }));
  };
  const handleDownloadDataToolbar = () => {
    buildAndDownloadCsv(data);
  };
  return hasControls ? /* @__PURE__ */ jsx(MapInitialViewProvider, { children: /* @__PURE__ */ jsxs(
    ChartToolbarRenderer,
    {
      visible: mouseInBounds,
      collapsed: toolbarConfig?.collapsed,
      draggable: toolbarConfig?.draggable,
      placement: toolbarConfig?.placement ?? defaultToolbarPlacement,
      handleDownloadDataToolbar,
      isDownloadEnabled: !hasOnlyDownloadCSV && isDownloadDataEnabled,
      initialA11yState: initialA11yToolbarTabIndexContext,
      onMouseEnter: handleMouseEnter,
      onMouseLeave: handleMouseLeave,
      menuDisplayMode: "never",
      children: [
        isZoomEnabled && /* @__PURE__ */ jsx(MapZoomInOutButtons, {}),
        isZoomEnabled && /* @__PURE__ */ jsx(MapResetButton, {}),
        isZoomToFitEnabled && /* @__PURE__ */ jsx(MapZoomToFitButton, {}),
        isDownloadDataEnabled && /* @__PURE__ */ jsx(MapDownloadDataButton, {})
      ]
    }
  ) }) : null;
};
MapToolbar["displayName"] = "MapToolbar";
export {
  MapToolbar
};
