// @dynatrace/strato-components/esm/charts/core/components/overlay-tooltip/components/series/SeriesItemContent.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { assignInlineVars } from "@vanilla-extract/dynamic";
import clsx from "clsx";
import { isEmpty } from "lodash-es";
import { forwardRef } from "react";
import { SeriesSymbol } from "./SeriesSymbol.js";
import { Tooltip } from "../../../../../../overlays/tooltip/Tooltip.js";
import { GAP_BETWEEN_COLUMNS } from "../../constants.js";
import {
  multiLineName,
  relativeValueStyle,
  seriesDescriptionText,
  seriesDescriptionTextMultiLine,
  seriesDescriptionTextMultiLineWrap,
  seriesItem,
  seriesItemMaxWidthVar,
  seriesItemText,
  seriesNameCell,
  seriesNameText,
  seriesValueCell,
  spaceSavedForSymbolVar,
  symbolWrapper
} from "../../tooltip.sty.js";
const SeriesItemContent = forwardRef((props, ref) => {
  const {
    color,
    nameText,
    descriptionText,
    fullDescriptionText,
    isDescriptionTruncated,
    descriptionShouldUseMultiLineCss,
    descriptionMultiLineMode = "clamp",
    displayValueText,
    displayRelativeValueText,
    fullValueText,
    fullRelativeValueText,
    nameTextStyle,
    nameStyle,
    multiLine = false,
    symbolPosition,
    shape,
    symbol,
    icon,
    iconStyle,
    maxWidth,
    hasChevron = false
  } = props;
  const isRightPosition = symbolPosition === "right";
  const hasSymbol = color !== void 0 && !isEmpty(color) || icon !== void 0 || symbol !== void 0;
  const renderNameContent = () => {
    if (Array.isArray(nameText) && multiLine) {
      return /* @__PURE__ */ jsx("ul", { className: multiLineName, children: nameText.map((line) => {
        return /* @__PURE__ */ jsx(
          "li",
          {
            className: nameTextStyle,
            style: { ...nameStyle, padding: 0 },
            children: /* @__PURE__ */ jsx("span", { className: seriesNameText, children: line })
          },
          `${line}`
        );
      }) });
    }
    return /* @__PURE__ */ jsx("span", { className: nameTextStyle, style: nameStyle, children: /* @__PURE__ */ jsx("span", { className: seriesNameText, children: nameText }) });
  };
  const renderSymbol = () => {
    return /* @__PURE__ */ jsx("span", { className: symbolWrapper({ isRight: isRightPosition }), children: /* @__PURE__ */ jsx(
      SeriesSymbol,
      {
        color,
        shape,
        symbol,
        icon,
        iconStyle
      }
    ) });
  };
  const getDescriptionClassName = () => {
    if (!descriptionShouldUseMultiLineCss) {
      return seriesDescriptionText;
    }
    return descriptionMultiLineMode === "wrap" ? seriesDescriptionTextMultiLineWrap : seriesDescriptionTextMultiLine;
  };
  const renderValue = () => {
    const isValueTruncated = displayValueText !== fullValueText;
    const isRelativeValueTruncated = displayRelativeValueText !== fullRelativeValueText;
    return /* @__PURE__ */ jsxs(Fragment, { children: [
      /* @__PURE__ */ jsx(Tooltip, { text: fullValueText, disabled: !isValueTruncated, children: /* @__PURE__ */ jsx("span", { className: seriesItemText, children: displayValueText }) }),
      displayRelativeValueText && /* @__PURE__ */ jsx(
        Tooltip,
        {
          text: fullRelativeValueText,
          disabled: !isRelativeValueTruncated,
          children: /* @__PURE__ */ jsx("span", { className: clsx(relativeValueStyle, seriesItemText), children: displayRelativeValueText })
        }
      )
    ] });
  };
  return /* @__PURE__ */ jsxs(
    "div",
    {
      ref,
      className: seriesItem({ symbolPosition, hasChevron }),
      style: assignInlineVars({
        [seriesItemMaxWidthVar]: `${maxWidth}px`,
        [spaceSavedForSymbolVar]: hasSymbol ? "24px" : "6px"
      }),
      children: [
        /* @__PURE__ */ jsx("div", { style: { gridArea: "symbol" }, children: hasSymbol ? renderSymbol() : null }),
        /* @__PURE__ */ jsxs(
          "div",
          {
            className: seriesNameCell,
            style: {
              gridArea: "name",
              marginRight: `${GAP_BETWEEN_COLUMNS}px`,
              flexDirection: "column"
            },
            children: [
              renderNameContent(),
              descriptionText !== void 0 && /* @__PURE__ */ jsx(
                Tooltip,
                {
                  text: fullDescriptionText ?? "",
                  disabled: !isDescriptionTruncated,
                  children: /* @__PURE__ */ jsx(
                    "span",
                    {
                      className: getDescriptionClassName(),
                      style: descriptionShouldUseMultiLineCss ? nameStyle : void 0,
                      children: descriptionText
                    }
                  )
                }
              )
            ]
          }
        ),
        /* @__PURE__ */ jsx("div", { className: seriesValueCell, style: { gridArea: "value" }, children: renderValue() })
      ]
    }
  );
});
export {
  SeriesItemContent
};
