// @dynatrace/strato-components/esm/charts/core/components/overlay-tooltip/components/series/SeriesItem.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import { noop } from "lodash-es";
import { forwardRef, useRef } from "react";
import { useIntl } from "react-intl";
import { ChevronRightSmallIcon } from "@dynatrace/strato-icons";
import { useSeriesItemDisplayState } from "./hooks/use-series-item-display-state.js";
import { SeriesActionsHeader } from "./SeriesActionsHeader.js";
import { SeriesItemContent } from "./SeriesItemContent.js";
import { useId } from "../../../../../../core/hooks/useId.js";
import { COPY_TO_CLIPBOARD_MESSAGES } from "../../../copy-to-clipboard/copy-to-clipboard-messages.js";
import { ChartActionsMenuAsync } from "../../../series-action/ChartActionsMenuAsync.js";
import { TOOLTIP_CONTENT_PADDING } from "../../constants.js";
import { useExternalConfig } from "../../contexts/ExternalConfigContext.js";
import { useDefaultLabels } from "../../hooks/useDefaultLabels.js";
import { useSelector, useTooltipSnapshot } from "../../selectors.js";
import {
  chevron,
  chevronRightAligned,
  seriesItemActive,
  seriesItemWithChevron,
  seriesItemWrapper,
  TOOLTIP_MAX_WIDTH
} from "../../tooltip.sty.js";
import { useTooltipReducer } from "../../useTooltipReducer.js";
import { getDefaultAction } from "../../utils/get-copy-action.js";
const ACTIVE_SERIES_HOVER_DELAY = 300;
const CHEVRON_SLOT_WIDTH = 20;
const SeriesItem = (props) => {
  const intl = useIntl();
  const { pinned } = useTooltipSnapshot();
  const labelFor = useDefaultLabels();
  const { actions, ...restProps } = props;
  const { symbolPosition = "left", defaultActionsConfig } = useExternalConfig();
  const hasActions = actions !== void 0 || defaultActionsConfig.size > 0;
  const dispatch = useTooltipReducer();
  const rowId = useId();
  const isOpen = useSelector(({ activeSeries: activeSeries2 }) => activeSeries2 === rowId);
  const { activeSeries } = useTooltipSnapshot();
  const menuDelayRef = useRef(0);
  const asyncActions = actions ? async () => actions() : void 0;
  const triggerLeftTooltip = (event) => {
    event.stopPropagation();
    dispatch({ type: "TOOLTIP_MOUSE_LEAVE" })();
  };
  const handleClick = (event) => {
    event.stopPropagation();
    menuDelayRef.current = 0;
    dispatch({
      type: "SET_ACTIVE_SERIES",
      payload: { activeSeries: rowId }
    })();
  };
  const handlePointerMove = () => {
    menuDelayRef.current = ACTIVE_SERIES_HOVER_DELAY;
    dispatch({
      type: "SET_ACTIVE_SERIES",
      payload: { activeSeries: rowId }
    })();
  };
  const defaultActions = !pinned ? null : [...defaultActionsConfig.entries()].map(([accessor, config]) => {
    const message = intl.formatMessage(
      config.messageLabel ?? COPY_TO_CLIPBOARD_MESSAGES.copyField,
      {
        field: labelFor?.(
          config.field,
          config.customLabel
        )?.toLocaleLowerCase()
      }
    );
    return getDefaultAction(props, accessor, message);
  });
  return /* @__PURE__ */ jsx(
    "button",
    {
      className: clsx(
        activeSeries === rowId && seriesItemActive,
        seriesItemWrapper
      ),
      onClick: handleClick,
      onPointerMove: handlePointerMove,
      onKeyDown: noop,
      "data-testid": "menu-trigger",
      children: /* @__PURE__ */ jsx(
        ChartActionsMenuAsync,
        {
          hasActions,
          defaultActions,
          actions: asyncActions,
          isOpen,
          header: /* @__PURE__ */ jsx(SeriesActionsHeader, { text: props.name }),
          onActionSelected: triggerLeftTooltip,
          delay: menuDelayRef.current,
          children: /* @__PURE__ */ jsx(
            WithSymbol,
            {
              symbolPosition,
              hasActions,
              ...restProps
            }
          )
        }
      )
    }
  );
};
const WithSymbol = forwardRef(
  (props, ref) => {
    const containerWidth = TOOLTIP_MAX_WIDTH - TOOLTIP_CONTENT_PADDING;
    const {
      color,
      multiLine = false,
      name,
      description,
      shape,
      iconStyle,
      icon,
      symbol
    } = props;
    const { symbolPosition, hasActions } = props;
    const seriesContentMaxWidth = hasActions ? containerWidth - CHEVRON_SLOT_WIDTH : containerWidth;
    const {
      nameText,
      descriptionText,
      fullDescriptionText,
      isDescriptionTruncated,
      descriptionShouldUseMultiLineCss,
      descriptionMultiLineMode,
      displayValueText,
      displayRelativeValueText,
      fullValueText,
      fullRelativeValueText,
      nameTextStyle,
      nameStyle
    } = useSeriesItemDisplayState({
      ...props,
      maxWidth: seriesContentMaxWidth
    });
    return /* @__PURE__ */ jsxs("div", { ref, className: seriesItemWithChevron, children: [
      /* @__PURE__ */ jsx(
        SeriesItemContent,
        {
          color,
          name,
          description,
          nameText,
          descriptionText,
          fullDescriptionText,
          isDescriptionTruncated,
          descriptionShouldUseMultiLineCss,
          descriptionMultiLineMode,
          displayValueText,
          displayRelativeValueText,
          fullValueText,
          fullRelativeValueText,
          nameTextStyle,
          nameStyle,
          multiLine,
          symbolPosition,
          shape,
          symbol,
          iconStyle,
          icon,
          hasChevron: hasActions,
          maxWidth: seriesContentMaxWidth
        }
      ),
      hasActions && /* @__PURE__ */ jsx(
        "span",
        {
          className: clsx(
            chevron,
            symbolPosition === "right" && chevronRightAligned
          ),
          "aria-hidden": true,
          children: /* @__PURE__ */ jsx(ChevronRightSmallIcon, { size: 16 })
        }
      )
    ] });
  }
);
export {
  SeriesItem
};
