// @dynatrace/strato-components/esm/charts/core/components/overlay-tooltip/components/series/Series.js
import { jsx, jsxs } from "react/jsx-runtime";
import { useVirtualizer } from "@tanstack/react-virtual";
import clsx from "clsx";
import { Fragment, useRef, useCallback } from "react";
import {
  MIN_TOOLTIP_WIDTH,
  SCROLLBAR_WIDTH,
  SYMBOL_AND_SPACING_WIDTH,
  TOOLTIP_CONTENT_PADDING,
  TOOLTIP_WRAPPER_PADDING
} from "../../constants.js";
import { useExternalConfig } from "../../contexts/ExternalConfigContext.js";
import { useScrollbarDetection } from "../../hooks/use-scrollbar-detection.js";
import { useSelector } from "../../selectors.js";
import {
  seriesGridContainer,
  seriesItemHighlighted,
  seriesOverflowContainer,
  seriesWrapper
} from "../../tooltip.sty.js";
import { estimateSeriesItemSize } from "../../utils/estimate-series-item-size.js";
import { calculateTotalWidth400 } from "../../utils/width-calculation.js";
const Series = (props) => {
  const { data, seriesItemTemplate, activeItem = 0 } = props;
  const config = useExternalConfig();
  if (config.variant !== "shared" || data.length === 1) {
    return /* @__PURE__ */ jsxs("div", { className: seriesGridContainer, children: [
      /* @__PURE__ */ jsx(Fragment, { children: config.seriesHeader }, "seriesHeader"),
      /* @__PURE__ */ jsx(Fragment, { children: seriesItemTemplate(data[0]) }, "series")
    ] });
  }
  const maxContentWidthWithinFirstTenValues = Math.max(
    ...data.slice(0, 10).map((item) => {
      let nameWidth = 0;
      if (Array.isArray(item.name)) {
        nameWidth = calculateTotalWidth400(...item.name);
      } else {
        nameWidth = calculateTotalWidth400(item.name);
      }
      let valueWidth = 0;
      if (item.value !== void 0 && item.value !== null) {
        const valueText = String(item.value);
        const relativeValueText = item.relativeValue ? String(item.relativeValue) : "";
        valueWidth = calculateTotalWidth400(valueText, relativeValueText);
      }
      return nameWidth + valueWidth;
    })
  );
  return /* @__PURE__ */ jsx(
    SeriesWithVirtualizer,
    {
      data,
      activeItem,
      seriesItemTemplate,
      maxContentWidth: maxContentWidthWithinFirstTenValues
    }
  );
};
const SeriesWithVirtualizer = (props) => {
  const { data, seriesItemTemplate, activeItem, maxContentWidth } = props;
  const isItemSelected = useSelector(({ activeSeries }) => !!activeSeries);
  const parentRef = useRef(null);
  const config = useExternalConfig();
  const isSeriesHeaderDisabled = config.seriesHeader?.props?.disabled;
  const scrollbarRefCallback = useScrollbarDetection();
  const combinedRefCallback = useCallback(
    (element) => {
      parentRef.current = element;
      scrollbarRefCallback(element);
    },
    [scrollbarRefCallback]
  );
  const virtualizer = useVirtualizer({
    count: data.length,
    getScrollElement: () => parentRef.current,
    estimateSize: (index) => estimateSeriesItemSize(data[index], maxContainerWidth),
    paddingStart: isSeriesHeaderDisabled ? 0 : 24,
    paddingEnd: 0,
    overscan: 3
  });
  const spacings = TOOLTIP_WRAPPER_PADDING * 2 + SYMBOL_AND_SPACING_WIDTH;
  const maxContainerWidth = maxContentWidth + spacings + TOOLTIP_CONTENT_PADDING * 2;
  return /* @__PURE__ */ jsx("div", { ref: combinedRefCallback, className: seriesWrapper, children: /* @__PURE__ */ jsxs(
    "div",
    {
      className: seriesOverflowContainer,
      style: {
        height: `${virtualizer.getTotalSize()}px`,
        width: `${Math.max(maxContainerWidth, MIN_TOOLTIP_WIDTH)}px`,
        maxWidth: "100%"
      },
      children: [
        config.seriesHeader,
        virtualizer.getVirtualItems().map((item) => {
          return /* @__PURE__ */ jsx(
            "div",
            {
              className: clsx(
                seriesGridContainer,
                item.index === activeItem && !isItemSelected && seriesItemHighlighted
              ),
              style: {
                position: "absolute",
                top: 0,
                left: 0,
                width: `calc(100% - var(--needs-scrollbar-fix, 0) * ${SCROLLBAR_WIDTH}px)`,
                height: `${item.size}px`,
                transform: `translateY(${item.start}px)`
              },
              children: seriesItemTemplate(data[item.index])
            },
            item.key
          );
        })
      ]
    }
  ) });
};
export {
  Series
};
