// @dynatrace/strato-components/charts/meter-bar/shared/constants.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var constants_exports = {};
__export(constants_exports, {
  DEFAULT_FORMATTER: () => DEFAULT_FORMATTER,
  DEFAULT_GAP_WIDTH: () => DEFAULT_GAP_WIDTH,
  DEFAULT_HEIGHT: () => DEFAULT_HEIGHT,
  DEFAULT_WIDTH: () => DEFAULT_WIDTH,
  INITIAL_STATE: () => INITIAL_STATE,
  LARGE_CHART_SIZE: () => LARGE_CHART_SIZE,
  MAX_DEFAULT_VALUE: () => MAX_DEFAULT_VALUE,
  MEDIUM_CHART_SIZE: () => MEDIUM_CHART_SIZE,
  METER_BAR_FONT_COLOR: () => METER_BAR_FONT_COLOR,
  METER_BAR_MIN_MAX_FONT_COLOR: () => METER_BAR_MIN_MAX_FONT_COLOR,
  METER_BAR_SIZE: () => METER_BAR_SIZE,
  MIN_DEFAULT_VALUE: () => MIN_DEFAULT_VALUE,
  SIZE_DEFAULT: () => SIZE_DEFAULT,
  SMALL_CHART_SIZE: () => SMALL_CHART_SIZE
});
module.exports = __toCommonJS(constants_exports);
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
const INITIAL_STATE = Object.freeze({
  position: {
    absolute: { x: 0, y: 0 },
    relative: { x: 0, y: 0 }
  },
  pinned: false,
  visible: false
});
const METER_BAR_SIZE = Object.freeze({
  Size16: "size16",
  Size24: "size24",
  Size8: "size8",
  Auto: "auto"
});
const MIN_DEFAULT_VALUE = 0;
const MAX_DEFAULT_VALUE = 100;
const SIZE_DEFAULT = METER_BAR_SIZE.Size16;
const DEFAULT_FORMATTER = (v) => v.toString();
const DEFAULT_GAP_WIDTH = 1;
const DEFAULT_WIDTH = "100%";
const DEFAULT_HEIGHT = "100%";
const SMALL_CHART_SIZE = 8;
const MEDIUM_CHART_SIZE = 16;
const LARGE_CHART_SIZE = 24;
const METER_BAR_FONT_COLOR = import_colors.default.Text.Neutral.Default;
const METER_BAR_MIN_MAX_FONT_COLOR = import_colors.default.Text.Neutral.Subdued;
