// @dynatrace/strato-components/charts/histogram/components/displays/utils/datapoint-stacking.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var datapoint_stacking_exports = {};
__export(datapoint_stacking_exports, {
  getCumulativeValue: () => getCumulativeValue,
  getStackedKeyByUnit: () => getStackedKeyByUnit,
  setCumulativeBins: () => setCumulativeBins
});
module.exports = __toCommonJS(datapoint_stacking_exports);
function getStackedKeyByUnit(bin, unit) {
  return bin.from.toString() + Math.sign(bin.value).toString() + bin.from.toString() + bin.to.toString() + //TODO: Pending to apply hash function from this PR: https://bitbucket.lab.dynatrace.org/projects/APPFW/repos/design-system/pull-requests/7188/overview?commentId=1144690
  (unit ?? "").toString();
}
function getCumulativeValue(mapStackedPoints, bin, unit, seriesId, value = 0) {
  const key = getStackedKeyByUnit(bin, unit);
  const stackedBin = mapStackedPoints.get(key);
  let binsCumulative = 0;
  let baseCumulative = 0;
  let cumulativeValue = value;
  if (stackedBin) {
    const {
      binsCumulative: prevBinsCumulative,
      baseCumulative: prevBaseCumulative,
      seriesId: prevSeriesId
    } = stackedBin;
    binsCumulative = prevBinsCumulative ?? 0;
    baseCumulative = prevBaseCumulative ?? 0;
    const baseCumulativeHasToChange = prevSeriesId !== seriesId;
    if (baseCumulativeHasToChange) {
      baseCumulative = binsCumulative;
    }
    cumulativeValue = baseCumulative + value;
  }
  binsCumulative = stackedBin?.seriesId !== seriesId ? cumulativeValue : Math.sign(value) > 0 ? Math.max(cumulativeValue, binsCumulative) : Math.min(cumulativeValue, binsCumulative);
  return {
    cumulativeValue,
    binsCumulative,
    baseCumulative,
    key
  };
}
function setCumulativeBins(mapStackedPoints, seriesId, key, binsCumulative, baseCumulative) {
  mapStackedPoints.set(key, {
    seriesId,
    binsCumulative,
    baseCumulative
  });
}
