// @dynatrace/strato-components/esm/filters/filter-field/FilterField.js
import { jsx, jsxs } from "react/jsx-runtime";
import { startCompletion } from "@codemirror/autocomplete";
import { syntaxTree } from "@codemirror/language";
import { EditorState, Transaction } from "@codemirror/state";
import {
  EditorView,
  placeholder as placeholderExtension
} from "@codemirror/view";
import clsx from "clsx";
import { AnimatePresence } from "framer-motion";
import { isEqual } from "lodash-es";
import {
  Children,
  forwardRef,
  useCallback,
  useContext,
  useEffect,
  useImperativeHandle,
  useMemo,
  useReducer,
  useRef,
  useState
} from "react";
import { ErrorBoundary, useErrorBoundary } from "react-error-boundary";
import { useIntl } from "react-intl";
import { FilterIcon } from "@dynatrace/strato-icons";
import { FilterFieldSuffixButtons } from "./components/FilterFieldSuffixButtons.js";
import { FilterFieldContext } from "./contexts/FilterFieldContext.js";
import { FilterFieldLoggerContext } from "./contexts/FilterFieldLoggerContext.js";
import { handleEditMode } from "./editor-state-configs/edit-mode-update-listener.js";
import {
  ANNOTATION_COMPOSE_END,
  ANNOTATION_DELETE_STATEMENT,
  ANNOTATION_DOC_CHANGED_CONTROLLED,
  ANNOTATION_END_PASTING_MODE,
  ANNOTATION_SUGGESTION_APPLY,
  disabledPropCompartment,
  overrideKeys,
  pastingEffect,
  placeholderCompartment,
  readOnlyPropCompartment,
  stringAnnotation,
  suggestionsCompartment,
  validatorMapEffect
} from "./editor-state-configs/editor-config.js";
import {
  handleControlledPasting,
  handlePastingTransaction
} from "./editor-state-configs/pasting-update-listener.js";
import { ErrorFallback } from "./error-fallback/ErrorFallback.js";
import { obfuscateFilterString } from "./error-fallback/filter-value-obfuscation-util.js";
import { DynamicFilterFieldConfigurations } from "./extensions/compartments.js";
import { FilterFieldInteractionsExtension } from "./extensions/interactions-extensions.js";
import { FilterFieldLanguageExtension } from "./extensions/language-extension.js";
import { FilterFieldRecentSavedFiltersExtension } from "./extensions/recent-saved-filters.js";
import { FilterFieldSyntaxHighlightingExtension } from "./extensions/syntax-highlighting-extension.js";
import * as styles from "./FilterField.sty.js";
import { usePinnedFilterSuggestions } from "./hooks/recent-and-pinned/usePinnedFilterSuggestions.js";
import { useRecentFilterMutation } from "./hooks/recent-and-pinned/useRecentFilters.js";
import { useRecentFilterSuggestions } from "./hooks/recent-and-pinned/useRecentFilterSuggestions.js";
import { useCachedTreeConversion } from "./hooks/useCachedTreeConversion.js";
import { useErrorAwareViewRef } from "./hooks/useErrorAwareViewRef.js";
import { useErrorLogsStore } from "./hooks/useErrorLogsStore.js";
import { useFilterFieldLinting } from "./hooks/useFilterFieldLinting.js";
import { useSuggestions } from "./hooks/useSuggestions.js";
import { useUpdatableCallback } from "./hooks/useUpdatableCallback.js";
import { filterFieldMessages } from "./messages.js";
import { getNumberParser } from "./parser/getNumberParser.js";
import { filterFieldReducer } from "./state/filter-field-state-reducer.js";
import { Suggestion } from "./suggestions/Suggestion.js";
import { SuggestionDetails } from "./suggestions/SuggestionDetails.js";
import { SuggestionGroup } from "./suggestions/SuggestionGroup.js";
import { SuggestionGroupLabel } from "./suggestions/SuggestionGroupLabel.js";
import { DEFAULT_SUGGESTIONS_COUNT } from "./suggestions/suggestions-config.js";
import { suggestionsExtension } from "./suggestions/suggestions-extension.js";
import { Suggestions } from "./suggestions/Suggestions.js";
import { SuggestionsOverlay } from "./suggestions/SuggestionsOverlay.js";
import { autoFormatLogicalOperators } from "./utils/auto-format-logical-operators.js";
import {
  getCurrentFocusableListItem,
  getFirstFocusableListItem,
  getLastFocusableListItem,
  getNextFocusableListItem
} from "./utils/get-next-suggestion.js";
import {
  toInternalValidatorMap
} from "./utils/validation-utils/normalize-validator-map.js";
import { VisuallyHidden } from "../../core/components/visually-hidden/VisuallyHidden.js";
import { OverlayStackingContext } from "../../core/contexts/OverlayStacking.js";
import { useAriaLabelingProps as _useAriaLabelingProps } from "../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useFormControlState } from "../../core/hooks/useFormControlState.js";
import { useId } from "../../core/hooks/useId.js";
import { useMergeRefs } from "../../core/hooks/useMergeRefs.js";
import { createStratoLogger } from "../../core/sdk/sdk-logger.js";
import { isSlot } from "../../core/utils/is-slot.js";
import * as editorStyles from "../../editors/base-code-editor/BaseCodeEditor.sty.js";
import { withFormFieldMessageContextProvider } from "../../forms/forms-core/components/form-field/FormFieldMessageContextProvider.js";
import { InternalBaseItem as BaseItem } from "../../forms/forms-core/components/form-field-messages/FormFieldMessageBaseItem.js";
import { useFormFieldMessagesContext } from "../../forms/forms-core/hooks/useFormFieldMessagesContext.js";
import { getI18nErrorMessage } from "../../forms/forms-core/hooks/useFormFieldValidation.js";
import { useSyncedFormControlProps } from "../../forms/forms-core/hooks/useSyncedFormControlProps.js";
import { FormFieldMessagesAction } from "../../forms/forms-core/state/formfield-messages-state-reducer.js";
import { useInputHint } from "../../forms/hooks/useInputHint.js";
import { Flex } from "../../layouts/flex/Flex.js";
import { closeAllOverlays } from "../../overlays/hooks/useOverlay.js";
import { useOverlayWithTrigger } from "../../overlays/hooks/useOverlayWithTrigger.js";
import { Text } from "../../typography/text/Text.js";
const COMPONENT_NAME = "FilterField";
function isEditorSelection(selection) {
  return selection.ranges !== void 0;
}
const _FilterField = forwardRef(
  (props, forwardedRef) => {
    const {
      children,
      name,
      value,
      defaultValue,
      required: propRequired,
      disabled: propDisabled,
      readOnly: propReadOnly = false,
      placeholder: placeholderProp,
      id: propId,
      autoSuggestions = false,
      defaultSuggestionsCount: _defaultSuggestionsCount,
      controlState,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "aria-disabled": propAriaDisabled,
      onChange,
      onSuggest,
      onFilter,
      validatorMap,
      parserConfig,
      filterNamespace,
      customTypes,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps] = useBehavioralTrackingProps(remainingProps);
    const setCursorPosition = useErrorLogsStore(
      (state2) => state2.setCursorPosition
    );
    const setAction = useErrorLogsStore((state2) => state2.setAction);
    const setIsControlled = useErrorLogsStore((state2) => state2.setIsControlled);
    const suggestionsSlot = Children.toArray(children).find(
      (child) => isSlot(child, Suggestions)
    );
    const { getSyntaxTree } = useCachedTreeConversion();
    const intl = useIntl();
    const { showBoundary } = useErrorBoundary();
    const onChangeRef = useUpdatableCallback(onChange);
    const onFilterRef = useUpdatableCallback(onFilter);
    const validatorMapRef = useRef(validatorMap);
    const customTypesRef = useRef(customTypes);
    const internalValidatorMapRef = useRef();
    const numberParser = getNumberParser();
    const isUserEventComposeRef = useRef(false);
    const valuePlaceholderWildcardSelection = useRef({ isActive: false });
    const {
      id,
      disabled,
      required,
      ariaDisabled,
      readOnly: _readOnly
    } = useSyncedFormControlProps({
      propAriaDisabled,
      propDisabled,
      propId,
      propRequired,
      propReadOnly
    });
    const readOnly = !!_readOnly;
    const [prevValue, setPrevValue] = useState(value);
    const [prevPropDisabled, setPrevPropDisabled] = useState(disabled);
    const [prevPropRequired, setPrevPropRequired] = useState(required);
    const [prevPropReadOnly, setPrevPropReadOnly] = useState(readOnly);
    const [prevPropId, setPrevPropId] = useState(id);
    const [prevPlaceholderProp, setPrevPlaceholderProp] = useState(placeholderProp);
    const [prevOnSuggest, setPrevOnSuggest] = useState(() => onSuggest);
    const isControlled = value !== void 0;
    useEffect(() => {
      setIsControlled(isControlled);
    }, [isControlled, setIsControlled]);
    const placeholder = intl.formatMessage(filterFieldMessages.placeholder);
    const overlayStacking = useContext(OverlayStackingContext);
    const suggestionsOverlayRef = useRef(null);
    const {
      targetRef,
      overlayRef: _overlayRef,
      overlayState,
      overlayProps: _overlayProps
    } = useOverlayWithTrigger({
      placement: "bottom-start",
      placementFallbacks: ["top-start"],
      widthStrategy: "trigger",
      offset: 4,
      // Workaround to close all other overlays when opening the suggestions overlay to avoid overlapping overlays.
      // TODO: Refactor with APPDEV-3373.
      onOpenChange(isOpen) {
        if (isOpen && !suggestionsOverlayRef.current && editorViewRef.current?.hasFocus) {
          closeAllOverlays(overlayStacking);
          requestAnimationFrame(() => {
            editorViewRef.current?.focus();
          });
        }
      }
    });
    const overlayRef = useMergeRefs([
      suggestionsOverlayRef,
      _overlayRef
    ]);
    const { style: overlayStyles } = _overlayProps;
    const triggerRef = useRef(null);
    const editorViewRef = useRef(null);
    const internalRootRef = useRef(null);
    const rootRef = useMergeRefs([
      targetRef,
      internalRootRef
    ]);
    const hiddenInputRef = useRef(null);
    const ariaLabelingProps = _useAriaLabelingProps(props);
    const pastingInfoRef = useRef({
      pastingInProgress: false
    });
    const { viewDispatch } = useErrorAwareViewRef(editorViewRef);
    const filterFieldLinter = useFilterFieldLinting(customTypes);
    const [hasFormControlError, hint] = useFormControlState(controlState);
    const [hasSyntaxError, setHasSyntaxError] = useState(false);
    const {
      state: formFieldMessageState,
      dispatch: formFieldMessageContextDispatch
    } = useFormFieldMessagesContext();
    const [showError, setShowError] = useState(false);
    useEffect(() => {
      if (showError && !hasFormControlError && !hasSyntaxError && !formFieldMessageState.showError) {
        setShowError(false);
      }
      if (!editorViewRef.current?.hasFocus && formFieldMessageState.isTouched && (hasFormControlError || formFieldMessageState.showError || hasSyntaxError)) {
        setShowError(true);
      }
    }, [
      hasFormControlError,
      hasSyntaxError,
      formFieldMessageState.showError,
      editorViewRef.current?.hasFocus
    ]);
    const { inputProps, hintProps } = useInputHint({
      hint,
      hasError: showError
    });
    const isDirtyRef = useRef(false);
    const [prevDisabled, setPrevDisabled] = useState(disabled);
    if (disabled !== prevDisabled) {
      setPrevDisabled(disabled);
      viewDispatch({
        effects: disabledPropCompartment.reconfigure(
          EditorView.editable.of(!disabled)
        )
      });
    }
    const [prevReadOnly, setPrevReadOnly] = useState(readOnly);
    if (readOnly !== prevReadOnly) {
      setPrevReadOnly(readOnly);
      viewDispatch({
        effects: readOnlyPropCompartment.reconfigure(
          EditorState.readOnly.of(readOnly || false)
        )
      });
    }
    const filterFieldCustomErrorRef = useRef(useId());
    const filterFieldRequiredErrorRef = useRef(useId());
    const validate = useCallback(() => {
      if (valueRef.current === "" && required) {
        formFieldMessageContextDispatch?.({
          type: FormFieldMessagesAction.ADDMESSAGE,
          payload: {
            id: filterFieldRequiredErrorRef.current,
            validityStateKey: "valueMissing",
            variant: "error",
            message: /* @__PURE__ */ jsx(Text, { textStyle: "small", as: "p", children: getI18nErrorMessage("valueMissing", hiddenInputRef.current) }),
            messageType: "formcontrol"
          }
        });
      } else if (valueRef.current !== "" && required) {
        formFieldMessageContextDispatch?.({
          type: FormFieldMessagesAction.REMOVEMESSAGE,
          payload: {
            id: filterFieldRequiredErrorRef.current,
            messageType: "formcontrol"
          }
        });
      }
    }, [formFieldMessageContextDispatch, required]);
    const updateSyntaxError = useCallback(
      (isValid) => {
        setHasSyntaxError(!isValid);
        if (!isValid) {
          formFieldMessageContextDispatch?.({
            type: FormFieldMessagesAction.ADDMESSAGE,
            payload: {
              id: filterFieldCustomErrorRef.current,
              validityStateKey: "customError",
              variant: "error",
              message: intl.formatMessage(filterFieldMessages.syntaxError),
              messageType: "formcontrol"
            }
          });
        } else {
          formFieldMessageContextDispatch?.({
            type: FormFieldMessagesAction.REMOVEMESSAGE,
            payload: {
              id: filterFieldCustomErrorRef.current,
              messageType: "formcontrol"
            }
          });
        }
      },
      [formFieldMessageContextDispatch, setHasSyntaxError, intl]
    );
    const onInputChange = () => {
      if (isDirtyRef.current) {
        validate();
      }
    };
    const onInputBlur = () => {
      isDirtyRef.current = true;
      validate();
      formFieldMessageContextDispatch?.({
        type: FormFieldMessagesAction.SETISTOUCHED,
        payload: { isTouched: true }
      });
    };
    const saveRecentFilter = useRecentFilterMutation(filterNamespace);
    useRecentFilterSuggestions(editorViewRef, filterNamespace);
    usePinnedFilterSuggestions(
      editorViewRef,
      filterNamespace,
      overlayState.isOpen
    );
    useImperativeHandle(
      forwardedRef,
      () => ({
        element: internalRootRef.current,
        triggerRef: triggerRef.current,
        inputRef: hiddenInputRef.current,
        focus: () => {
          hiddenInputRef.current?.focus();
          editorViewRef.current?.dispatch({
            selection: { anchor: editorViewRef.current.state.doc.length }
          });
        },
        open: () => hiddenInputRef.current?.focus(),
        close: () => overlayState.close(),
        validate: () => {
          validate();
          formFieldMessageContextDispatch?.({
            type: FormFieldMessagesAction.SETISTOUCHED,
            payload: { isTouched: true }
          });
        },
        saveRecentFilter: async () => {
          if (filterNamespace) {
            await saveRecentFilter(valueRef.current).catch(() => void 0);
          }
        }
      }),
      [
        overlayState,
        validate,
        formFieldMessageContextDispatch,
        filterNamespace,
        saveRecentFilter
      ]
    );
    const [state, dispatch] = useReducer(filterFieldReducer, {
      suggestions: /* @__PURE__ */ new Map(),
      activeSuggestion: -1
    });
    const [hideDeleteWhileTyping, setHideDeleteWhileTyping] = useState(false);
    const {
      ariaAttributes,
      autoSuggestionsConfig,
      recentSuggestionValues,
      pinnedSuggestionValues,
      logicalOperatorSuggestionValues,
      pastedContentSuggestionValues,
      searchSuggestionValues,
      fallbackKeySuggestionValues,
      statementSuggestionValues,
      focusedListItem,
      suggestionsContext,
      applySuggestion,
      handleNextSuggestion,
      suggestionsOverlayHintId,
      _onSuggest
    } = useSuggestions({
      ariaDisabled,
      autoSuggestions,
      isOpen: overlayState.isOpen,
      id,
      suggestionsOverlayRef,
      editorViewRef,
      onSuggest,
      viewDispatch,
      hiddenInputRef,
      ariaLabelingProps
    });
    const valueRef = useRef(value ?? defaultValue ?? "");
    const transactionRef = useRef(void 0);
    const isValidSearchRef = useRef(false);
    useEffect(() => {
      if (!triggerRef.current) {
        return;
      }
      function handleOnChange(transaction) {
        if (!editorViewRef.current) {
          return transaction;
        }
        let newValue = transaction.newDoc.toString();
        let transactions = [];
        const tree = syntaxTree(transaction.state);
        const {
          tree: convertedTree,
          isValid: isValid2,
          isValidSearch: isValidSearch2
        } = getSyntaxTree({
          tree,
          view: editorViewRef.current,
          numberParser,
          doc: transaction.newDoc.toString(),
          internalValidatorMap: internalValidatorMapRef.current,
          customTypes: customTypesRef.current,
          parserConfig
        });
        isValidSearchRef.current = isValidSearch2;
        updateSyntaxError(isValid2);
        onInputChange();
        const annotation = transaction.annotation(stringAnnotation);
        if (annotation === ANNOTATION_DOC_CHANGED_CONTROLLED) {
          return transaction;
        }
        if (annotation !== ANNOTATION_SUGGESTION_APPLY) {
          setAction({
            type: "handleOnChange",
            data: {
              value: valueRef.current,
              newValue
            }
          });
        }
        if (newValue !== valueRef.current) {
          if (isControlled) {
            handleControlledPasting(transaction, pastingInfoRef);
          }
          setHideDeleteWhileTyping(true);
          if (!transaction.isUserEvent("delete.backward") && !transaction.isUserEvent("delete.forward")) {
            [newValue, transactions] = autoFormatLogicalOperators(
              newValue,
              tree,
              transaction.selection?.ranges[transaction.selection?.mainIndex ?? 0].to ?? transaction.newDoc.toString().length
            );
          }
          onChangeRef.current?.(newValue, convertedTree, isValid2);
          transactionRef.current = [{ selection: transaction.newSelection }];
          const userEvent = transaction.annotation(Transaction.userEvent);
          transactionRef.current = [
            {
              ...transaction,
              // On Windows / Linux, hitting 'Shift + Delete' results in an undefined selection,
              // leading to the suggestions callback not being triggered.
              // Even for an empty value, there should always be a selection.
              selection: transaction.newSelection ?? { anchor: 0 },
              // Annotations are not automatically stored, we need to do that manually
              // to be able to determine, which action triggered the transaction when applying
              // the new value in a controlled scenario.
              annotations: Transaction.userEvent.of(
                userEvent ? userEvent : transaction.annotation(stringAnnotation) ?? ""
              )
            },
            ...transactions
          ];
          if (!isControlled) {
            valueRef.current = newValue;
            onInputChange();
          }
          return isControlled ? [] : [transaction, ...transactions];
        }
        return [transaction];
      }
      internalValidatorMapRef.current = toInternalValidatorMap(validatorMap);
      const initialEditorState = EditorState.create({
        doc: value ?? defaultValue ?? "",
        extensions: [
          FilterFieldLanguageExtension,
          filterFieldLinter,
          FilterFieldSyntaxHighlightingExtension,
          FilterFieldInteractionsExtension({
            closeSuggestions: overlayState.close,
            showBoundary
          }),
          DynamicFilterFieldConfigurations({
            ariaAttributes,
            disabled: !!disabled,
            readOnly,
            placeholder: placeholderProp ?? placeholder
          }),
          EditorView.domEventObservers({
            compositionend: () => {
              isUserEventComposeRef.current = false;
              editorViewRef.current?.dispatch({
                annotations: stringAnnotation.of(ANNOTATION_COMPOSE_END)
              });
            }
          }),
          FilterFieldRecentSavedFiltersExtension,
          /**
           * Suggestions
           */
          suggestionsCompartment.of(
            suggestionsExtension({
              overlayState,
              numberParser,
              onSuggest,
              internalValidatorMap: internalValidatorMapRef.current,
              _onSuggest,
              getSyntaxTree,
              parserConfig,
              customTypes: customTypesRef.current
            })
          ),
          /**
           * Listeners
           */
          // Update cursor position for error logging whenever it changes.
          EditorState.transactionFilter.of((transaction) => {
            const userEvent = transaction.annotation(Transaction.userEvent);
            if (userEvent?.includes("select")) {
              setCursorPosition(transaction.state.selection.main.anchor);
            }
            return transaction;
          }),
          // Intercept transactions that change the doc to not instantly apply the changes
          // in a controlled scenario. Instead, the transaction is stored so we can apply
          // it if the consumer actually updates the value.
          EditorState.transactionFilter.of((transaction) => {
            if (!editorViewRef.current) {
              return transaction;
            }
            const annotation = transaction.annotation(stringAnnotation);
            if (annotation === ANNOTATION_DOC_CHANGED_CONTROLLED) {
              return handleOnChange(transaction);
            }
            if (annotation === ANNOTATION_END_PASTING_MODE) {
              queueMicrotask(() => {
                try {
                  if (editorViewRef?.current) {
                    startCompletion(editorViewRef.current);
                  }
                } catch (e) {
                  showBoundary(e);
                }
              });
            }
            let userEvent = transaction.annotation(Transaction.userEvent);
            if (userEvent === void 0) {
              switch (annotation) {
                case ANNOTATION_COMPOSE_END:
                  userEvent = ANNOTATION_COMPOSE_END;
                  break;
                case ANNOTATION_DELETE_STATEMENT:
                  userEvent = ANNOTATION_DELETE_STATEMENT;
                  break;
                case ANNOTATION_SUGGESTION_APPLY: {
                  userEvent = ANNOTATION_SUGGESTION_APPLY;
                  const selection = transaction.selection?.ranges;
                  if (selection && selection[0] && selection[0].to - selection[0].from === 5) {
                    valuePlaceholderWildcardSelection.current = {
                      isActive: true,
                      range: { from: selection[0].from, to: selection[0].to },
                      potentialNewValue: transaction.newDoc.toString()
                    };
                  }
                  break;
                }
              }
            }
            if (userEvent === "input.type.compose.start") {
              isUserEventComposeRef.current = true;
            }
            if (!userEvent || isUserEventComposeRef.current) {
              return transaction;
            }
            if ((transaction.docChanged && userEvent || !isUserEventComposeRef.current) && // We add the space character in the updateListener, so we don't want to call OnChange.
            !transaction.isUserEvent("select")) {
              return handleOnChange(transaction);
            }
            return transaction;
          }),
          EditorView.updateListener.of((update) => {
            const isUndoOrRedo = update.transactions.some(
              (transaction) => transaction.isUserEvent("undo") || transaction.isUserEvent("redo")
            );
            if (update.docChanged && isUndoOrRedo) {
              update.transactions.forEach(
                (transaction) => handleOnChange(transaction)
              );
              editorView?.dispatch({
                effects: [
                  pastingEffect.of({
                    pastingInProgress: false
                  })
                ]
              });
            }
          }),
          // Update Listener for pasting
          EditorView.updateListener.of((update) => {
            handlePastingTransaction(
              update,
              pastingInfoRef,
              editorView,
              editorViewRef
            );
          }),
          // Automatically insert a space at the end of the input when entering edit mode,
          // if the cursor is at the end, the statement is complete, and no trailing space or bracket is present.
          // This starts a new statement.
          EditorView.updateListener.of((update) => {
            handleEditMode(update, suggestionsContext);
          })
        ]
      });
      const editorView = new EditorView({
        state: initialEditorState,
        parent: triggerRef.current
      });
      editorViewRef.current = editorView;
      if (validatorMap) {
        editorView.dispatch({
          effects: [validatorMapEffect.of(validatorMapRef?.current)],
          // In order to re-run the decorations and re-generate errors and diagnostics,
          // we have to dispatch a new document.
          changes: {
            from: 0,
            to: editorViewRef.current.state.doc.length,
            insert: editorViewRef.current.state.doc
          }
        });
      }
      const { isValid, isValidSearch } = getSyntaxTree({
        tree: syntaxTree(editorView.state),
        view: editorView,
        numberParser,
        doc: editorView.state.doc.toString(),
        internalValidatorMap: internalValidatorMapRef.current,
        customTypes: customTypesRef.current,
        parserConfig
      });
      isValidSearchRef.current = isValidSearch;
      updateSyntaxError(isValid);
      return () => editorView.destroy();
    }, []);
    const validatorMapChanged = !isEqual(validatorMap, validatorMapRef.current);
    const customTypesChanged = !isEqual(customTypes, customTypesRef.current);
    if ((validatorMapChanged || customTypesChanged) && editorViewRef.current) {
      if (validatorMapChanged) {
        validatorMapRef.current = validatorMap;
        internalValidatorMapRef.current = validatorMap ? toInternalValidatorMap(validatorMap) : void 0;
      }
      if (customTypesChanged) {
        customTypesRef.current = customTypes;
      }
      if (validatorMapChanged || customTypesChanged) {
        viewDispatch({
          effects: suggestionsCompartment.reconfigure(
            suggestionsExtension({
              overlayState,
              numberParser,
              onSuggest,
              _onSuggest,
              getSyntaxTree,
              internalValidatorMap: internalValidatorMapRef.current,
              customTypes: customTypesRef.current,
              parserConfig
            })
          )
        });
      }
      const doc = editorViewRef.current.state.doc;
      queueMicrotask(() => {
        try {
          if (editorViewRef.current) {
            const tree = syntaxTree(editorViewRef.current.state);
            const { tree: convertedTree, isValid } = getSyntaxTree({
              tree,
              view: editorViewRef.current,
              numberParser,
              doc: doc.toString(),
              internalValidatorMap: internalValidatorMapRef.current,
              customTypes: customTypesRef.current,
              parserConfig
            });
            updateSyntaxError(isValid);
            onChangeRef.current?.(doc.toString(), convertedTree, isValid);
            onInputChange();
            viewDispatch({
              effects: [validatorMapEffect.of(validatorMapRef?.current)],
              // In order to re-run the decorations and re-generate errors and diagnostics,
              // we have to dispatch a new document.
              changes: {
                from: 0,
                to: editorViewRef.current.state.doc.length,
                insert: doc.toString()
              }
            });
            if (validatorMapChanged || customTypesChanged) {
              queueMicrotask(() => {
                try {
                  if (editorViewRef.current) {
                    startCompletion(editorViewRef.current);
                  }
                } catch (e) {
                  showBoundary(e);
                }
              });
            }
          }
        } catch (e) {
          showBoundary(e);
        }
      });
    }
    if (placeholderProp !== prevPlaceholderProp) {
      viewDispatch({
        effects: placeholderCompartment.reconfigure(
          placeholderExtension(placeholderProp ?? placeholder)
        )
      });
      setPrevPlaceholderProp(placeholderProp);
    }
    if (onSuggest !== prevOnSuggest) {
      viewDispatch({
        effects: suggestionsCompartment.reconfigure(
          suggestionsExtension({
            overlayState,
            numberParser,
            onSuggest,
            _onSuggest,
            getSyntaxTree,
            internalValidatorMap: internalValidatorMapRef.current,
            customTypes: customTypesRef.current,
            parserConfig
          })
        )
      });
      setPrevOnSuggest(() => onSuggest);
    }
    if (value !== prevValue && isControlled) {
      setPrevValue(value);
      valueRef.current = value;
      queueMicrotask(() => {
        try {
          setAction({
            type: "controlledValueChange",
            data: {
              value: prevValue ?? "",
              newValue: valueRef.current
            }
          });
          onInputChange();
          const lastTransaction = transactionRef.current && transactionRef.current[transactionRef.current.length - 1];
          const selection = lastTransaction?.selection ?? // Fall back to the current editor selection in case the last transaction is undefined
          // to avoid a jumping cursor when typing really fast.
          editorViewRef.current?.state.selection;
          const anchor = Math.min(
            value.length,
            (isEditorSelection(selection) ? selection.ranges[0].anchor : selection?.anchor) ?? value.length
          );
          let dispatchedSelection = {
            anchor
          };
          if (valuePlaceholderWildcardSelection.current?.isActive && valuePlaceholderWildcardSelection.current?.potentialNewValue === value) {
            const selectionFrom = valuePlaceholderWildcardSelection.current.range?.from;
            const selectionTo = valuePlaceholderWildcardSelection.current.range?.to;
            if (selectionFrom && selectionFrom <= value.length && selectionTo && selectionTo <= value.length) {
              dispatchedSelection = {
                anchor: selectionFrom,
                head: selectionTo
              };
            }
          }
          viewDispatch({
            changes: {
              from: 0,
              to: editorViewRef.current?.state.doc.length ?? 0,
              insert: value
            },
            annotations: stringAnnotation.of(ANNOTATION_DOC_CHANGED_CONTROLLED),
            selection: dispatchedSelection
          });
          if (valuePlaceholderWildcardSelection.current) {
            valuePlaceholderWildcardSelection.current = {
              isActive: false
            };
          }
          transactionRef.current = void 0;
        } catch (e) {
          showBoundary(e);
        }
      });
    }
    if (disabled !== prevPropDisabled || required !== prevPropRequired || readOnly !== prevPropReadOnly || id !== prevPropId) {
      setPrevPropDisabled(disabled);
      setPrevPropRequired(required);
      setPrevPropReadOnly(readOnly);
      setPrevPropId(id);
      if (disabled !== prevPropDisabled) {
        viewDispatch({
          effects: disabledPropCompartment.reconfigure(
            EditorView.editable.of(!disabled)
          )
        });
      }
      if (readOnly !== prevPropReadOnly) {
        viewDispatch({
          effects: readOnlyPropCompartment.reconfigure(
            EditorState.readOnly.of(readOnly)
          )
        });
      }
    }
    const [hasAriaLivePolite, setHasAriaLivePolite] = useState(false);
    useEffect(() => {
      setHasAriaLivePolite(false);
      const debounceTimer = setTimeout(() => {
        setHasAriaLivePolite(true);
      }, 300);
      return () => {
        clearTimeout(debounceTimer);
      };
    }, [valueRef.current]);
    const defaultSuggestionsCount = useMemo(() => {
      if (typeof _defaultSuggestionsCount === "number") {
        return _defaultSuggestionsCount;
      }
      return {
        empty: _defaultSuggestionsCount?.empty ?? DEFAULT_SUGGESTIONS_COUNT.empty,
        filled: _defaultSuggestionsCount?.filled ?? DEFAULT_SUGGESTIONS_COUNT.filled
      };
    }, [_defaultSuggestionsCount]);
    const filterFieldContextValue = useMemo(
      () => ({
        hasAriaLivePolite,
        applySuggestion,
        dispatch,
        focusedListItem,
        filterFieldId: id,
        term: suggestionsContext.current?.token.textValue,
        autoSuggestions: autoSuggestions ? autoSuggestionsConfig : {},
        recentSuggestionValues,
        pinnedSuggestionValues,
        logicalOperatorSuggestionValues,
        pastedContentSuggestionValues,
        searchSuggestionValues,
        fallbackKeySuggestionValues,
        fallbackKeys: internalValidatorMapRef.current?.fallbackKeys,
        statementSuggestionValues,
        filterStatementKeysMap: internalValidatorMapRef.current?.filterStatementKeysMap,
        triggerRef: triggerRef.current ?? null,
        suggestionsOverlayRef: suggestionsOverlayRef.current ?? null,
        filterNamespace: filterNamespace ?? null,
        editorView: editorViewRef.current,
        parserConfig,
        isValidSearch: isValidSearchRef,
        defaultSuggestionsCount,
        handleNextSuggestion
      }),
      [
        hasAriaLivePolite,
        applySuggestion,
        focusedListItem,
        id,
        suggestionsContext,
        autoSuggestions,
        autoSuggestionsConfig,
        recentSuggestionValues,
        pinnedSuggestionValues,
        logicalOperatorSuggestionValues,
        pastedContentSuggestionValues,
        searchSuggestionValues,
        fallbackKeySuggestionValues,
        internalValidatorMapRef,
        statementSuggestionValues,
        filterNamespace,
        parserConfig,
        defaultSuggestionsCount,
        handleNextSuggestion
      ]
    );
    return /* @__PURE__ */ jsx(FilterFieldContext.Provider, { value: filterFieldContextValue, children: /* @__PURE__ */ jsxs(
      Flex,
      {
        ref: rootRef,
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        gap: 4,
        flexDirection: "column",
        className: clsx(styles.wrapper, consumerClassName),
        style: consumerStyle,
        onMouseMove: () => {
          setHideDeleteWhileTyping(false);
        },
        ...behavioralTrackingProps,
        "data-dt-component": COMPONENT_NAME,
        children: [
          /* @__PURE__ */ jsxs(
            "div",
            {
              ...inputProps,
              "aria-invalid": controlState ? inputProps["aria-invalid"] : hasSyntaxError,
              className: clsx(
                editorStyles.editorContainer({
                  size: "default",
                  fullHeight: false,
                  // check if the editor is disabled first to apply proper styling
                  hasError: !(disabled || ariaDisabled) && showError,
                  lineNumbers: "hidden",
                  readOnly: disabled || readOnly
                }),
                styles.container,
                props.className,
                {
                  [styles.errorText]: showError,
                  [styles.errorVars]: showError,
                  [styles.disabled]: disabled,
                  [styles.readOnly]: readOnly,
                  [styles.ariaDisabled]: ariaDisabled,
                  hideDeleteWhileTyping
                }
              ),
              onBlur: onInputBlur,
              onKeyDown: (event) => {
                if (event.target instanceof HTMLElement && event.target.closest("button")) {
                  return;
                }
                if (readOnly && event.key !== "Tab" && !event.ctrlKey) {
                  event.preventDefault();
                  return;
                }
                switch (event.key) {
                  case "Space":
                    if (!editorViewRef.current) {
                      return;
                    }
                    if (event.ctrlKey || event.altKey) {
                      overlayState.open();
                      startCompletion(editorViewRef.current);
                    }
                    break;
                  case "Enter": {
                    event.preventDefault();
                    if (event.shiftKey || !editorViewRef.current) {
                      return;
                    }
                    if (!focusedListItem || event.ctrlKey || event.metaKey) {
                      validate();
                      const { tree, isValid } = getSyntaxTree({
                        tree: syntaxTree(editorViewRef.current.state),
                        view: editorViewRef.current,
                        numberParser,
                        doc: editorViewRef.current.state.doc.toString(),
                        internalValidatorMap: internalValidatorMapRef.current,
                        customTypes,
                        parserConfig
                      });
                      onFilterRef.current?.({
                        value: valueRef.current,
                        syntaxTree: tree,
                        isValid
                      });
                      overlayState.close();
                      if (isValid && hiddenInputRef.current?.form) {
                        let inputCount = 0;
                        for (const element of Array.from(
                          hiddenInputRef.current.form.elements
                        )) {
                          if (!element.hasAttribute("hidden") && (element.tagName === "INPUT" && element.getAttribute("type") !== "hidden" || element.tagName === "SELECT" || element.tagName === "TEXTAREA")) {
                            inputCount += 1;
                            if (inputCount > 1) {
                              break;
                            }
                          }
                        }
                        if (inputCount === 1) {
                          hiddenInputRef.current?.form?.requestSubmit();
                        }
                      } else {
                        setTimeout(() => {
                          hiddenInputRef.current?.form?.reportValidity();
                        });
                      }
                    } else if (focusedListItem && focusedListItem.id) {
                      if (state.suggestions.has(focusedListItem.id)) {
                        state.suggestions.get(focusedListItem.id)?.();
                      } else {
                        const focusedElement = document.querySelector(
                          `[data-strato-filter-field-virtual-focus-id="${focusedListItem?.id}"]`
                        );
                        focusedElement?.click();
                      }
                      return;
                    }
                    break;
                  }
                }
                if (overlayState.isOpen) {
                  if (overrideKeys.includes(event.key)) {
                    event.preventDefault();
                    event.stopPropagation();
                  }
                  if (event.key === "ArrowUp" || event.key === "ArrowDown") {
                    const direction = event.key === "ArrowUp" ? "previous" : "next";
                    const overlay = suggestionsOverlayRef.current;
                    const currentlyFocusedItem = getCurrentFocusableListItem(
                      overlay,
                      focusedListItem?.id
                    );
                    const isAtEnd = currentlyFocusedItem === getLastFocusableListItem(overlay) && direction === "next";
                    const isAtStart = currentlyFocusedItem === getFirstFocusableListItem(overlay) && direction === "previous";
                    if (isAtEnd || isAtStart) {
                      const scrollContainer = overlay?.querySelector(
                        '[role="listbox"]'
                      )?.firstElementChild;
                      if (scrollContainer) {
                        scrollContainer.scroll({
                          top: direction === "next" ? 0 : scrollContainer.scrollHeight
                        });
                        requestAnimationFrame(() => {
                          const nextSuggestion2 = getNextFocusableListItem(
                            direction,
                            overlay,
                            focusedListItem?.id
                          );
                          handleNextSuggestion(nextSuggestion2);
                        });
                      }
                      return;
                    }
                    const nextSuggestion = getNextFocusableListItem(
                      direction,
                      overlay,
                      focusedListItem?.id
                    );
                    handleNextSuggestion(nextSuggestion);
                  }
                }
              },
              onKeyDownCapture: (event) => {
                if ((event.key === "Home" || event.key === "End" || event.key === "PageUp" || event.key === "PageDown") && focusedListItem) {
                  event.stopPropagation();
                  event.preventDefault();
                  let nextSuggestion = null;
                  if (event.key === "PageUp" || event.key === "PageDown") {
                    const direction = event.key === "PageUp" ? "previous" : "next";
                    nextSuggestion = getNextFocusableListItem(
                      direction,
                      suggestionsOverlayRef.current,
                      focusedListItem?.id,
                      5
                    );
                    handleNextSuggestion(nextSuggestion);
                  } else if (event.key === "Home" || event.key === "End") {
                    const overlay = suggestionsOverlayRef.current;
                    const scrollContainer = overlay?.querySelector(
                      '[role="listbox"]'
                    )?.firstElementChild;
                    if (scrollContainer) {
                      scrollContainer.scroll({
                        top: event.key === "Home" ? 0 : scrollContainer.scrollHeight
                      });
                    }
                    requestAnimationFrame(() => {
                      const nextSuggestion2 = event.key === "Home" ? getFirstFocusableListItem(overlay) : getLastFocusableListItem(overlay);
                      handleNextSuggestion(nextSuggestion2);
                    });
                  }
                }
              },
              onMouseDown: (event) => {
                if (disabled) {
                  event.preventDefault();
                }
              },
              children: [
                /* @__PURE__ */ jsx(
                  "span",
                  {
                    className: styles.filterIcon({
                      hasError: showError,
                      disabled: disabled || Boolean(ariaDisabled)
                    }),
                    children: /* @__PURE__ */ jsx(FilterIcon, {})
                  }
                ),
                /* @__PURE__ */ jsx("div", { ref: triggerRef, style: { maxHeight: "inherit" } }),
                /* @__PURE__ */ jsx(
                  FilterFieldSuffixButtons,
                  {
                    value: valueRef.current,
                    ariaDisabled,
                    disabled,
                    showError,
                    readOnly
                  }
                )
              ]
            }
          ),
          /* @__PURE__ */ jsx(AnimatePresence, { children: !readOnly && overlayState.isOpen && /* @__PURE__ */ jsx(
            SuggestionsOverlay,
            {
              ref: overlayRef,
              "data-testid": dataTestId ? `filterfield-suggestions-${dataTestId}` : void 0,
              overlayProps: { style: overlayStyles },
              "data-dtrum-mask": dataDtrumMask,
              "data-dtrum-allow": dataDtrumAllow,
              children: suggestionsSlot === void 0 ? /* @__PURE__ */ jsx(Suggestions, { children }) : children
            }
          ) }),
          hint && /* @__PURE__ */ jsx(
            BaseItem,
            {
              variant: showError ? "error" : "hint",
              disabled: disabled || ariaDisabled,
              ...hintProps,
              children: hint
            }
          ),
          /* @__PURE__ */ jsx(VisuallyHidden, { children: /* @__PURE__ */ jsx(
            "input",
            {
              ref: hiddenInputRef,
              name,
              value: valueRef.current,
              required,
              disabled,
              readOnly,
              tabIndex: -1,
              id,
              onFocus: () => {
                editorViewRef.current?.focus();
              },
              onChange: () => {
              },
              onInvalid: (event) => {
                event.preventDefault();
                setShowError(true);
                event.currentTarget.focus();
                onInputBlur();
              },
              "aria-hidden": "true"
            }
          ) }),
          /* @__PURE__ */ jsx(
            "p",
            {
              id: suggestionsOverlayHintId,
              className: styles.suggestionsOverlayAnnouncement,
              "aria-hidden": "true",
              children: intl.formatMessage(
                filterFieldMessages.suggestions.a11y.overlayAnnouncement
              )
            }
          )
        ]
      }
    ) });
  }
);
_FilterField.displayName = COMPONENT_NAME;
const InternalFilterField = withFormFieldMessageContextProvider(
  forwardRef((props, forwardedRef) => {
    const stratoLogger = useMemo(
      () => createStratoLogger.bind({
        name: props.name,
        id: props.id
      })(COMPONENT_NAME),
      [props.name, props.id]
    );
    const cursorPosition = useErrorLogsStore((state) => state.cursorPosition);
    const action = useErrorLogsStore((state) => state.action);
    const isControlled = useErrorLogsStore((state) => state.isControlled);
    const loggerContext = useMemo(
      () => ({ logger: stratoLogger }),
      [stratoLogger]
    );
    return /* @__PURE__ */ jsx(
      ErrorBoundary,
      {
        onError: (error) => {
          if (action !== null) {
            action.data.value = obfuscateFilterString(
              action.data.value,
              props.validatorMap,
              props.parserConfig
            );
            if (action.type === "handleOnChange" || action.type === "controlledValueChange") {
              action.data.newValue = obfuscateFilterString(
                action.data.newValue,
                props.validatorMap,
                props.parserConfig
              );
            }
          }
          stratoLogger.error(error?.message, error, {
            filterString: obfuscateFilterString(
              props.value ?? props.defaultValue ?? "",
              props.validatorMap,
              props.parserConfig
            ),
            cursorPosition,
            action,
            isControlled
          });
        },
        FallbackComponent: (fallbackProps) => /* @__PURE__ */ jsx(ErrorFallback, { ...fallbackProps }),
        children: /* @__PURE__ */ jsx(FilterFieldLoggerContext.Provider, { value: loggerContext, children: /* @__PURE__ */ jsx(_FilterField, { ...props, ref: forwardedRef }) })
      }
    );
  })
);
const FilterField = Object.assign(InternalFilterField, {
  Suggestions,
  Suggestion,
  SuggestionGroup,
  SuggestionGroupLabel,
  SuggestionDetails
});
export {
  FilterField
};
