// @dynatrace/strato-components/charts/timeseries/components/annotations-timeseries/TimeseriesChartAnnotationsArea.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TimeseriesChartAnnotationsArea_exports = {};
__export(TimeseriesChartAnnotationsArea_exports, {
  TimeseriesChartAnnotationsArea: () => TimeseriesChartAnnotationsArea
});
module.exports = __toCommonJS(TimeseriesChartAnnotationsArea_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_constants = require("./constants.js");
var import_useBehavioralTrackingProps = require("../../../../core/hooks/useBehavioralTrackingProps.js");
var import_AnnotationsTimeDisplay = require("../../../core/components/annotations/components/displays/AnnotationsTimeDisplay.js");
var import_AnnotationsScrollWrapper = require("../../../core/components/annotations/components/overlays/AnnotationsScrollWrapper.js");
var import_constants2 = require("../../../core/components/annotations/constants.js");
var import_ChartGridRoot = require("../../../core/components/chart-grid/ChartGridRoot.js");
var import_useGridPlotSize = require("../../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_selectors = require("../../hooks/store/selectors.js");
var import_useAnnotationsConfig = require("../../hooks/useAnnotationsConfig.js");
var import_useXScale = require("../../hooks/useXScale.js");
var import_timeseries_grid_area_rules = require("../../types/timeseries-grid-area-rules.js");
const COMPONENT_NAME = "AnnotationsChart";
const TimeseriesChartAnnotationsArea = () => {
  const config = (0, import_useAnnotationsConfig.useAnnotationsConfig)();
  const {
    tracks,
    showLabels = import_constants2.DEFAULT_SHOW_LABELS,
    annotationsActions,
    tooltip
  } = config;
  const [behaviorTrackingProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(config);
  const setState = (0, import_selectors.useSetState)();
  const { currentScale } = (0, import_useXScale.useXScales)();
  const scale = currentScale;
  const { width } = (0, import_useGridPlotSize.useGridPlotSize)();
  const handleUpdateIndicators = (event) => {
    setState((prev) => ({
      ...prev,
      annotations: {
        ...prev.annotations,
        ...event
      }
    }));
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ChartGridRoot.ChartGrid.GridArea,
    {
      paddingBottom: import_constants.ANNOTATIONS_PADDING_BOTTOM,
      gridArea: import_timeseries_grid_area_rules.AREAS.annotations,
      children: ({ availableWidth, availableHeight }) => {
        const labelOffset = availableWidth - width;
        return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_AnnotationsScrollWrapper.AnnotationsScrollWrapper,
          {
            height: availableHeight,
            width: availableWidth + import_constants2.SCROLLBAR_THRESHOLD,
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_AnnotationsTimeDisplay.AnnotationsTimeDisplay,
              {
                tracks,
                width: availableWidth,
                scale,
                showLabels: showLabels ?? import_constants2.DEFAULT_SHOW_LABELS,
                labelsOffset: labelOffset,
                annotationsActions: (annotation) => annotation && annotationsActions?.(annotation),
                markerGroupHover: handleUpdateIndicators,
                textOverflow: import_constants2.DEFAULT_INTEGRATION_TEXT_OVERFLOW,
                tooltipConfig: tooltip,
                ...behaviorTrackingProps,
                "data-dt-component": COMPONENT_NAME
              }
            )
          }
        );
      }
    }
  );
};
