// @dynatrace/strato-geo/esm/map/hooks/use-geo-tooltip-sync.js
import { isEqual } from "lodash-es";
import { useCallback, useEffect, useRef, useSyncExternalStore } from "react";
import {
  _useOverlayTooltipStore as useOverlayTooltipStore,
  _useOverlayTooltipReducer as useTooltipReducer
} from "@dynatrace/strato-components/charts";
import { buildGeoTooltipState } from "../utils/build-geo-tooltip-state.js";
const useOverlayTooltipSelector = (selector) => {
  const store = useOverlayTooltipStore();
  const prevRef = useRef(void 0);
  const getSnapshot = useCallback(() => {
    const next = selector(store.getState());
    if (prevRef.current !== void 0 && isEqual(prevRef.current, next)) {
      return prevRef.current;
    }
    prevRef.current = next;
    return next;
  }, [store, selector]);
  return useSyncExternalStore(store.subscribe, getSnapshot);
};
const useGeoOverlayTooltipState = () => {
  const metadata = useOverlayTooltipSelector(
    (state) => state.data?.metadata
  );
  const pinned = useOverlayTooltipSelector((state) => state.pinned);
  const inBounds = useOverlayTooltipSelector((state) => state.inBounds);
  return { data: metadata, pinned, inBounds };
};
const useGeoTooltipSync = (position, pinned, geometryType, geoData) => {
  const store = useOverlayTooltipStore();
  const dispatch = useTooltipReducer();
  useEffect(() => {
    const onScroll = () => {
      if (store.getState().pinned) {
        dispatch({ type: "TOOLTIP_MOUSE_LEAVE" })();
      }
    };
    document.body.addEventListener("scroll", onScroll, true);
    return () => {
      document.body.removeEventListener("scroll", onScroll, true);
    };
  }, [store, dispatch]);
  useEffect(() => {
    if (position) {
      const nextState = buildGeoTooltipState({
        absoluteX: position.x,
        absoluteY: position.y,
        relativeX: position.x,
        relativeY: position.y,
        pinned,
        geometry: geometryType,
        metadata: geoData
      });
      const current = store.getState();
      if (current.absoluteX !== nextState.absoluteX || current.absoluteY !== nextState.absoluteY || current.inBounds !== nextState.inBounds || current.pinned !== nextState.pinned || !isEqual(current.data?.metadata, nextState.data.metadata)) {
        store.setState(nextState);
      }
    } else {
      const currentState = store.getState();
      if (currentState.inBounds || currentState.pinned) {
        dispatch({ type: "SET_OUT_BOUNDS" })();
      }
    }
  }, [
    position?.x,
    position?.y,
    pinned,
    geometryType,
    geoData,
    store,
    dispatch
  ]);
};
export {
  useGeoOverlayTooltipState,
  useGeoTooltipSync,
  useOverlayTooltipSelector
};
