// @dynatrace/strato-geo/esm/map/hooks/use-overlay-events.js
import {
  _useOverlayTooltipReducer as useTooltipReducer,
  _useOverlayTooltipStore as useOverlayTooltipStore,
  _useOverlayChart as useOverlayChart
} from "@dynatrace/strato-components/charts";
import { useSetStateOverlay, useSetState } from "../store/store.js";
import { buildGeoTooltipState } from "../utils/build-geo-tooltip-state.js";
import { extractDataFromEvent } from "../utils/parse-tooltip-data.js";
const layerIdToGeometry = (layerId) => {
  if (layerId?.includes("bubble")) {
    return "geoBubble";
  }
  if (layerId?.includes("choropleth")) {
    return "geoChoropleth";
  }
  if (layerId?.includes("connection")) {
    return "geoConnection";
  }
  return "geoDot";
};
const useOverlayEvents = () => {
  const setOverlayState = useSetStateOverlay();
  const setState = useSetState();
  const dispatch = useTooltipReducer();
  const store = useOverlayTooltipStore();
  const overlay = useOverlayChart();
  const getAbsolutePosition = (event) => {
    return {
      x: Math.round(event.originalEvent.clientX),
      y: Math.round(event.originalEvent.clientY)
    };
  };
  const setTooltipMarker = (hoveredLayerId, location) => {
    setState((prev) => ({
      ...prev,
      tooltip: {
        ...prev.tooltip,
        hoveredLayerId,
        location
      }
    }));
  };
  const hideTooltip = () => {
    dispatch({ type: "RESET_TOOLTIP" })();
    setTooltipMarker(void 0, void 0);
  };
  const handleMouseEnter = () => {
    setOverlayState({ mouseInBounds: true });
  };
  const handleMouseMove = (event) => {
    if (event.features?.length) {
      const currentState = store.getState();
      if (currentState.pinned) {
        return;
      }
      const { data, hoveredLayerId } = extractDataFromEvent(event);
      overlay.clear();
      setTooltipMarker(hoveredLayerId, event.lngLat);
      const pos = getAbsolutePosition(event);
      store.setState(
        buildGeoTooltipState({
          absoluteX: pos.x,
          absoluteY: pos.y,
          relativeX: Math.round(event.point.x),
          relativeY: Math.round(event.point.y),
          pinned: false,
          geometry: layerIdToGeometry(hoveredLayerId),
          metadata: data
        })
      );
    } else {
      const currentState = store.getState();
      if (!currentState.pinned) {
        hideTooltip();
      }
    }
  };
  const handleMouseLeave = () => {
    setOverlayState({ mouseInBounds: false });
    const currentState = store.getState();
    if (!currentState.pinned) {
      hideTooltip();
    }
  };
  const handleMouseClick = (event) => {
    const { data, featureId, hoveredLayerId } = extractDataFromEvent(event);
    if (!featureId) {
      hideTooltip();
      return;
    }
    if (!hoveredLayerId) {
      hideTooltip();
      return;
    }
    overlay.clear();
    setTooltipMarker(hoveredLayerId, event.lngLat);
    const pos = getAbsolutePosition(event);
    store.setState(
      buildGeoTooltipState({
        absoluteX: pos.x,
        absoluteY: pos.y,
        relativeX: Math.round(event.point.x),
        relativeY: Math.round(event.point.y),
        pinned: true,
        geometry: layerIdToGeometry(hoveredLayerId),
        metadata: data
      })
    );
  };
  const hideTooltipIfNotPinned = () => {
    const currentState = store.getState();
    if (!currentState.pinned) {
      hideTooltip();
    }
  };
  return {
    handleMouseEnter,
    handleMouseMove,
    handleMouseLeave,
    handleMouseClick,
    handleZoom: hideTooltipIfNotPinned,
    handleDrag: hideTooltipIfNotPinned
  };
};
export {
  useOverlayEvents
};
