// @dynatrace/strato-components/buttons/button/Button.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Button_exports = {};
__export(Button_exports, {
  Button: () => Button,
  default: () => getButtonSlots
});
module.exports = __toCommonJS(Button_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_react_is = require("react-is");
var styles = __toESM(require("./Button.sty.js"));
var import_Label = require("./Label.js");
var import_Prefix = require("./Prefix.js");
var import_Suffix = require("./Suffix.js");
var import_ProgressCircle = require("../../content/progress/ProgressCircle.js");
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_focusRingStatic_css = require("../../core/styles/focusRingStatic.sty.js");
var import_merge_props = require("../../core/utils/merge-props.js");
var import_parse_boolean = require("../../core/utils/parse-boolean.js");
var import_Flex = require("../../layouts/flex/Flex.js");
var import_Grid = require("../../layouts/grid/Grid.js");
var import_field_css = require("../../styles/field.sty.js");
const COMPONENT_NAME = "Button";
function getButtonSlots(children) {
  const slots = {};
  import_react.Children.forEach(children, (child) => {
    if ((0, import_react_is.isElement)(child)) {
      switch (child.type) {
        case import_Suffix.Suffix:
          slots.suffix = child;
          break;
        case import_Prefix.Prefix:
          slots.prefix = child;
          break;
        case import_Label.Label:
          slots.label = child;
          break;
        default:
          slots.freeform ??= [];
          slots.freeform.push(child);
          break;
      }
    } else {
      if (child !== null && child !== void 0) {
        slots.freeform ??= [];
        slots.freeform.push(child);
      }
    }
  });
  return slots;
}
const _Button = /* @__PURE__ */ (0, import_react.forwardRef)(
  (props, ref) => {
    const {
      children,
      variant = "default",
      disabled,
      "aria-disabled": ariaDisabledProp,
      required,
      id: propId,
      type = "button",
      onClick,
      ariaLabel,
      width = "content",
      color = "neutral",
      textAlign = "center",
      size = "default",
      loading = false,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      as,
      ...remainingProps
    } = props;
    const slots = getButtonSlots(children);
    const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
    const ariaDisabled = (0, import_parse_boolean.parseBoolean)(ariaDisabledProp);
    const nativeDisabledOrLoading = (0, import_react.useMemo)(
      () => loading ? true : disabled,
      [disabled, loading]
    );
    const disabledOrLoading = (0, import_react.useMemo)(
      () => nativeDisabledOrLoading || ariaDisabled,
      [ariaDisabled, nativeDisabledOrLoading]
    );
    const ariaLabelingProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(restProps);
    const handleButtonClick = (event) => {
      if (nativeDisabledOrLoading) {
        event.preventDefault();
        event.stopPropagation();
        return;
      }
      onClick?.(event);
    };
    const buttonProps = {
      ref,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      required,
      disabled: nativeDisabledOrLoading && !ariaDisabled,
      id: propId,
      type,
      ...ariaLabelingProps,
      "aria-disabled": ariaDisabledProp,
      onClick: handleButtonClick
    };
    const disabledPropagationProps = {
      onPointerDown(evt) {
        if (nativeDisabledOrLoading) {
          evt.preventDefault();
          evt.stopPropagation();
        }
      },
      onPointerUp(evt) {
        if (nativeDisabledOrLoading) {
          evt.preventDefault();
          evt.stopPropagation();
        }
      }
    };
    const hasChildren = children || children === 0;
    const buttonMergedProps = {
      ...(0, import_merge_props.mergeProps)(
        buttonProps,
        disabledPropagationProps,
        // Need to cast this one as the inference of mergeProps would generate
        // a not allowed overlap. The restProps type is too complex due to the
        // polymorphic inheritance.
        restProps,
        behavioralTrackingProps
      )
    };
    const intl = (0, import_react_intl.useIntl)();
    const Component = as || "button";
    if (!hasChildren) {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        Component,
        {
          "data-dt-component": COMPONENT_NAME,
          ...buttonMergedProps,
          className: (0, import_clsx.default)(
            consumerClassName,
            "strato-inputgroup",
            (0, import_field_css.fieldTokens)({ variant, color, disabled: disabledOrLoading }),
            styles.button({
              disabled: disabledOrLoading,
              textAlign,
              size
            }),
            (0, import_focusRingStatic_css.focusRingStatic)({
              when: "focus-visible",
              color,
              disabled: nativeDisabledOrLoading
            }),
            {
              [`${styles.widths.content}`]: width === "content",
              [`${styles.widths.full}`]: width === "full"
            }
          ),
          style: {
            width: width !== "content" && width !== "full" ? width : void 0
          }
        }
      );
    }
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      Component,
      {
        "data-dt-component": COMPONENT_NAME,
        className: (0, import_clsx.default)(
          consumerClassName,
          "strato-inputgroup",
          (0, import_field_css.fieldTokens)({
            variant,
            color,
            disabled: disabledOrLoading || ariaDisabled
          }),
          styles.button({
            disabled: disabledOrLoading || ariaDisabled,
            textAlign,
            size
          }),
          {
            [`${styles.widths.content}`]: width === "content",
            [`${styles.widths.full}`]: width === "full"
          },
          (0, import_focusRingStatic_css.focusRingStatic)({
            when: "focus-visible",
            color,
            disabled: nativeDisabledOrLoading
          })
        ),
        style: {
          width: width !== "content" && width !== "full" ? width : void 0,
          ...consumerStyle
        },
        ...buttonMergedProps,
        children: [
          loading && /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: styles.loading, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_ProgressCircle.ProgressCircle,
            {
              size: "small",
              color,
              "aria-label": intl.formatMessage({
                defaultMessage: "Loading",
                id: "cycz+bYG+VH7Of4e",
                description: "Label for the loading state."
              })
            }
          ) }),
          /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
            import_Grid.Grid,
            {
              gap: size === "default" ? 4 : 2,
              grid: "auto / auto-flow auto",
              width: "100%",
              maxWidth: "100%",
              alignItems: "center",
              justifyContent: textAlign === "start" ? "space-between" : "center",
              style: {
                visibility: loading ? "hidden" : void 0
              },
              children: [
                textAlign === "start" ? /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { minWidth: 0, gap: size === "default" ? 4 : 2, children: [
                  slots.prefix,
                  slots.label,
                  slots.freeform && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(Button.Label, { children: slots.freeform })
                ] }) : /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
                  slots.prefix,
                  slots.label,
                  slots.freeform && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(Button.Label, { children: slots.freeform })
                ] }),
                slots.suffix
              ]
            }
          )
        ]
      }
    );
  }
);
_Button.displayName = COMPONENT_NAME;
const Button = Object.assign(_Button, {
  Suffix: import_Suffix.Suffix,
  Prefix: import_Prefix.Prefix,
  Label: import_Label.Label
});
