// @dynatrace/strato-components/tables/DataTable/hooks/useTable/createTable.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var createTable_exports = {};
__export(createTable_exports, {
  createTable: () => createTable
});
module.exports = __toCommonJS(createTable_exports);
var import_is_node_environment = require("../../../../core/utils/is-node-environment.js");
var import_ColumnFontStyle = require("../../features/ColumnFontStyle/ColumnFontStyle.js");
var import_ColumnOrder = require("../../features/ColumnOrder/ColumnOrder.js");
var import_ColumnPinning = require("../../features/ColumnPinning/ColumnPinning.js");
var import_ColumnSizing = require("../../features/ColumnSizing/ColumnSizing.js");
var import_ColumnVisibility = require("../../features/ColumnVisibility/ColumnVisibility.js");
var import_Download = require("../../features/Download/Download.js");
var import_GanttChart = require("../../features/GanttChart/GanttChart.js");
var import_Layout = require("../../features/Layout/Layout.js");
var import_LineWrap = require("../../features/LineWrap/LineWrap.js");
var import_Pagination = require("../../features/Pagination/Pagination.js");
var import_RowDetails = require("../../features/RowDetails/RowDetails.js");
var import_RowInteractivity = require("../../features/RowInteractivity/RowInteractivity.js");
var import_RowOrder = require("../../features/RowOrder/RowOrder.js");
var import_RowSelection = require("../../features/RowSelection/RowSelection.js");
var import_Sorting = require("../../features/Sorting/Sorting.js");
var import_SubRows = require("../../features/SubRows/SubRows.js");
var import_Thresholds = require("../../features/Thresholds/Thresholds.js");
var import_UserActions = require("../../features/UserActions/UserActions.js");
var import_VirtualScrolling = require("../../features/VirtualScrolling/VirtualScrolling.js");
var import_VisualVariants = require("../../features/VisualVariants/VisualVariants.js");
var import_src = require("../../tanstack-table-core-override/src/index.js");
const features = [
  import_src.Headers,
  import_src.ColumnVisibility,
  import_src.ColumnOrdering,
  import_src.ColumnPinning,
  import_src.ColumnFaceting,
  import_src.ColumnFiltering,
  import_src.GlobalFaceting,
  //depends on ColumnFaceting
  import_src.GlobalFiltering,
  //depends on ColumnFiltering
  import_src.RowSorting,
  import_src.ColumnGrouping,
  import_src.RowExpanding,
  import_src.RowPagination,
  import_src.RowSelection,
  import_src.ColumnSizing,
  import_ColumnSizing.DataTableColumnSizing,
  import_LineWrap.LineWrap,
  import_Layout.Layout,
  import_VisualVariants.VisualVariantConfig,
  import_UserActions.UserActions,
  import_Sorting.Sorting,
  import_ColumnVisibility.DataTableColumnVisibility,
  import_ColumnOrder.DataTableColumnOrder,
  import_RowSelection.DataTableRowSelection,
  import_Thresholds.Thresholds,
  import_SubRows.DataTableSubRows,
  import_RowDetails.DataTableRowDetails,
  import_RowInteractivity.DataTableRowInteractivity,
  import_VirtualScrolling.DataTableVirtualScrolling,
  import_Download.DataTableDownloadData,
  import_ColumnFontStyle.DataTableColumnFontStyle,
  import_Pagination.DataTablePaginationFeature,
  import_RowOrder.DataTableRowOrder,
  import_ColumnPinning.DataTableColumnPinning,
  import_GanttChart.DataTableGanttChartFeature
];
function createTable(options) {
  if (options.debugAll || options.debugTable) {
    console.info("Creating Table Instance...");
  }
  const table = { _features: features };
  const defaultOptions = table._features.reduce((obj, feature) => {
    return Object.assign(obj, feature.getDefaultOptions?.(table));
  }, {});
  const mergeOptions = (options2) => {
    if (table.options.mergeOptions) {
      return table.options.mergeOptions(defaultOptions, options2);
    }
    return {
      ...defaultOptions,
      ...options2
    };
  };
  const coreInitialState = {};
  let initialState = {
    ...coreInitialState,
    ...options.initialState ?? {}
  };
  table._features.forEach((feature) => {
    initialState = feature.getInitialState?.(initialState) ?? initialState;
  });
  const queued = [];
  let queuedTimeout = false;
  const coreInstance = {
    _features: features,
    options: {
      ...defaultOptions,
      ...options
    },
    initialState,
    _queue: (cb) => {
      queued.push(cb);
      if (!queuedTimeout) {
        queuedTimeout = true;
        Promise.resolve().then(() => {
          while (queued.length) {
            queued.shift()?.();
          }
          queuedTimeout = false;
        }).catch(
          (error) => setTimeout(() => {
            throw error;
          })
        );
      }
    },
    reset: () => {
      table.setState(table.initialState);
    },
    setOptions: (updater) => {
      const newOptions = (0, import_src.functionalUpdate)(updater, table.options);
      table.options = mergeOptions(newOptions);
    },
    getState: () => {
      return table.options.state;
    },
    setState: (updater) => {
      table.options.onStateChange?.(updater);
    },
    _getRowId: (row, index, parent) => table.options.getRowId?.(row, index, parent) ?? `${parent ? [parent.id, index].join(".") : index}`,
    getCoreRowModel: () => {
      table._getCoreRowModel ??= table.options.getCoreRowModel(table);
      return table._getCoreRowModel();
    },
    // The final calls start at the bottom of the model,
    // row ordering, which then work their way up
    getRowModel: () => {
      return table.getRowDetailsRowModel();
    },
    getRow: (id, searchAll) => {
      const row = (searchAll ? table.getCoreRowModel() : table.getRowModel()).rowsById[id];
      if (!row) {
        if (!(0, import_is_node_environment.isProduction)()) {
          throw new Error(`getRow expected an ID, but got ${id}`);
        }
        throw new Error();
      }
      return row;
    },
    _getDefaultColumnDef: (0, import_src.memo)(
      () => [table.options.defaultColumn],
      (defaultColumn) => {
        defaultColumn = defaultColumn ?? {};
        return {
          header: (props) => {
            const resolvedColumnDef = props.header.column.columnDef;
            if (resolvedColumnDef.accessorKey) {
              return resolvedColumnDef.accessorKey;
            }
            if (resolvedColumnDef.accessorFn) {
              return resolvedColumnDef.id;
            }
            return null;
          },
          // footer: props => props.header.column.id,
          cell: (props) => props.renderValue()?.toString?.() ?? null,
          ...table._features.reduce((obj, feature) => {
            return Object.assign(obj, feature.getDefaultColumnDef?.());
          }, {}),
          ...defaultColumn
        };
      },
      {
        debug: () => table.options.debugAll ?? table.options.debugColumns,
        key: (0, import_is_node_environment.isDev)() && "getDefaultColumnDef"
      }
    ),
    _getColumnDefs: () => table.options.columns,
    getAllColumns: (0, import_src.memo)(
      () => [table._getColumnDefs()],
      (columnDefs) => {
        const recurseColumns = (columnDefs2, parent, depth = 0) => {
          return columnDefs2.map((columnDef) => {
            const column = (0, import_src.createColumn)(
              // eslint-disable-next-line @typescript-eslint/prefer-ts-expect-error, @typescript-eslint/ban-ts-comment
              // @ts-ignore Strict mode transition - needs to be investigated
              table,
              columnDef,
              depth,
              parent
            );
            const groupingColumnDef = columnDef;
            column.columns = groupingColumnDef.columns ? recurseColumns(groupingColumnDef.columns, column, depth + 1) : [];
            return column;
          });
        };
        return recurseColumns(columnDefs);
      },
      {
        key: (0, import_is_node_environment.isDev)() && "getAllColumns",
        debug: () => table.options.debugAll ?? table.options.debugColumns
      }
    ),
    getAllFlatColumns: (0, import_src.memo)(
      () => [table.getAllColumns()],
      (allColumns) => {
        return allColumns.flatMap((column) => {
          return column.getFlatColumns();
        });
      },
      {
        key: (0, import_is_node_environment.isDev)() && "getAllFlatColumns",
        debug: () => table.options.debugAll ?? table.options.debugColumns
      }
    ),
    getAllFlatColumnIdsWithoutBuiltin: (0, import_src.memo)(
      () => [table.getAllColumns()],
      () => {
        return table.getAllFlatColumns().filter((leafColumn) => !leafColumn.columnDef.meta?.builtin).map((column) => column.id);
      },
      {
        key: (0, import_is_node_environment.isDev)() && "getAllFlatColumnIdsWithoutBuiltin",
        debug: () => table.options.debugAll ?? table.options.debugColumns
      }
    ),
    _getAllFlatColumnsById: (0, import_src.memo)(
      () => [table.getAllFlatColumns()],
      (flatColumns) => {
        return flatColumns.reduce(
          (acc, column) => {
            acc[column.id] = column;
            return acc;
          },
          {}
        );
      },
      {
        key: (0, import_is_node_environment.isDev)() && "getAllFlatColumnsById",
        debug: () => table.options.debugAll ?? table.options.debugColumns
      }
    ),
    getAllLeafColumns: (0, import_src.memo)(
      () => [table.getAllColumns(), table._getOrderColumnsFn()],
      (allColumns, orderColumns) => {
        const leafColumns = allColumns.flatMap(
          (column) => column.getLeafColumns()
        );
        return orderColumns(leafColumns);
      },
      {
        key: (0, import_is_node_environment.isDev)() && "getAllLeafColumns",
        debug: () => table.options.debugAll ?? table.options.debugColumns
      }
    ),
    getAllLeafColumnsWithoutBuiltin: (0, import_src.memo)(
      () => [table.getAllLeafColumns()],
      (allLeafColumns) => {
        return allLeafColumns.filter(
          (column) => !column.columnDef.meta?.builtin
        );
      },
      {
        key: (0, import_is_node_environment.isDev)() && "getAllLeafColumnsWithoutBuiltin",
        debug: () => table.options.debugAll ?? table.options.debugColumns
      }
    ),
    getColumn: (columnId) => {
      const column = table._getAllFlatColumnsById()[columnId];
      if (!(0, import_is_node_environment.isProduction)() && !column) {
        console.warn(`[Table] Column with id '${columnId}' does not exist.`);
      }
      return column;
    },
    getHeader: (0, import_src.memo)(
      (headerId) => [headerId, table.getLeafHeaders()],
      (headerId, leafHeaders) => {
        const header = leafHeaders.find((header2) => header2.id === headerId);
        if (!(0, import_is_node_environment.isProduction)() && !header) {
          console.warn(`[Table] Header with id '${headerId}' does not exist.`);
        }
        return header;
      },
      {
        key: (0, import_is_node_environment.isDev)() && "getHeader",
        debug: () => table.options.debugAll ?? table.options.debugHeaders
      }
    )
  };
  Object.assign(table, coreInstance);
  for (const feature of table._features) {
    feature?.createTable?.(table);
  }
  return table;
}
