// @dynatrace/strato-components/esm/charts/meter-bar/single-meter-bar/MeterBarChart.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import { forwardRef, useRef } from "react";
import { MeterBarChartPlotArea } from "./components/MeterBarChartPlotArea.js";
import { MeterBarChartThresholdLegend } from "./components/MeterBarChartThresholdLegend/MeterBarChartThresholdLegend.js";
import { METER_BAR_CHART_ARIA_LABEL } from "./constants.js";
import { useMeterBarChartConfig } from "./hooks/useMeterBarChartConfig.js";
import { MeterBarChartInternalDataProviders } from "./providers/MeterBarChartInternalDataProviders.js";
import { MeterBarChartProviders } from "./providers/MeterBarChartProviders.js";
import { Label } from "./slots/Label.js";
import { Legend } from "./slots/Legend.js";
import { Max } from "./slots/Max.js";
import { Min } from "./slots/Min.js";
import { ThresholdIndicator } from "./slots/ThresholdIndicator.js";
import { ThresholdLegend } from "./slots/ThresholdLegend.js";
import { Tooltip } from "./slots/Tooltip.js";
import { Value } from "./slots/Value.js";
import { useBehavioralTrackingProps } from "../../../core/hooks/useBehavioralTrackingProps.js";
import { Intent } from "../../../core/slots/Intent/Intent.js";
import { defaultChartSize } from "../../../core/styles/defaultChartValues.sty.js";
import { DEFAULT_VALUE_ACCESSOR } from "../../core/constants.js";
import { ColorRule } from "../../core/slots/ColorRule.js";
import { ChartType } from "../../core/types/chart-types.js";
import { MeterBarChartFormattedValue } from "../shared/components/formatted-value/MeterBarChartFormattedValue.js";
import { MeterBarLayout } from "../shared/components/meter-bar-layout/MeterBarLayout.js";
import { Toolbar } from "../shared/slots/Toolbar.js";
import { getNumericValue } from "../shared/utils/get-numeric-value.js";
const COMPONENT_NAME = "SingleMeterBarChart";
const MeterBarChartRoot = forwardRef((props, forwardedRef) => {
  const {
    width,
    height,
    value,
    name,
    color,
    className,
    style,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "data-testid": dataTestId = "single-meter-bar-chart-root",
    prefixIcon,
    valueAccessor
  } = props;
  const [behavioralTrackingProps] = useBehavioralTrackingProps(props);
  const config = useMeterBarChartConfig(props);
  const {
    size,
    thresholds,
    thresholdsIndicators,
    loading,
    min,
    max,
    minSlot,
    maxSlot,
    valueSlot,
    labelSlot,
    tooltip,
    formatter,
    unit
  } = config;
  const meterBarChartRef = useRef(null);
  const hasLegend = thresholds.legend ?? thresholdsIndicators.legend;
  const consumerConfig = {
    type: ChartType.SingleMeterBar,
    thresholdsIndicators,
    size,
    min,
    max,
    formatter,
    unit,
    tooltip: !tooltip.hidden,
    loading
  };
  const displayValue = getNumericValue(
    value,
    valueAccessor ?? DEFAULT_VALUE_ACCESSOR
  );
  const internalDataProvidersProps = {
    value,
    displayValue,
    valueAccessor,
    name,
    color
  };
  return /* @__PURE__ */ jsx(MeterBarChartProviders, { config, seriesActions: props.seriesActions, children: /* @__PURE__ */ jsxs(
    MeterBarLayout,
    {
      ...behavioralTrackingProps,
      ref: meterBarChartRef,
      size,
      width,
      height,
      className: clsx(defaultChartSize, className),
      style,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "data-testid": dataTestId,
      "data-dt-component": COMPONENT_NAME,
      "aria-label": METER_BAR_CHART_ARIA_LABEL,
      consumerConfig,
      userRef: forwardedRef,
      children: [
        /* @__PURE__ */ jsx(
          MeterBarLayout.Providers,
          {
            providerName: MeterBarChartInternalDataProviders,
            providerProps: internalDataProvidersProps
          }
        ),
        prefixIcon && /* @__PURE__ */ jsx(MeterBarLayout.Icon, { children: prefixIcon }),
        labelSlot && /* @__PURE__ */ jsx(MeterBarLayout.Label, { children: labelSlot }),
        valueSlot && /* @__PURE__ */ jsx(MeterBarLayout.Value, { children: /* @__PURE__ */ jsx(
          MeterBarChartFormattedValue,
          {
            value: displayValue,
            total: max,
            auto: valueSlot?.auto,
            children: valueSlot?.value
          }
        ) }),
        /* @__PURE__ */ jsx(MeterBarLayout.Chart, { children: /* @__PURE__ */ jsx(MeterBarChartPlotArea, {}) }),
        minSlot && /* @__PURE__ */ jsx(MeterBarLayout.Min, { children: /* @__PURE__ */ jsx(MeterBarChartFormattedValue, { value: min, auto: minSlot?.auto, children: minSlot?.value }) }),
        maxSlot && /* @__PURE__ */ jsx(MeterBarLayout.Max, { children: /* @__PURE__ */ jsx(MeterBarChartFormattedValue, { value: max, auto: maxSlot?.auto, children: maxSlot?.value }) }),
        hasLegend && /* @__PURE__ */ jsx(MeterBarLayout.Legend, { children: /* @__PURE__ */ jsx(MeterBarChartThresholdLegend, {}) })
      ]
    }
  ) });
});
MeterBarChartRoot.displayName = "MeterBarChart";
const MeterBarChart = Object.assign(MeterBarChartRoot, {
  Value,
  Label,
  Min,
  Max,
  ThresholdIndicator,
  // eslint-disable-next-line @typescript-eslint/no-deprecated
  ThresholdLegend,
  Legend,
  Tooltip,
  ColorRule,
  Intent,
  Toolbar
});
export {
  MeterBarChart
};
