// @dynatrace/strato-components/esm/charts/core/components/legend/range/components/RangeLegendSegments/RangeLegendSegments.js
import { jsx } from "react/jsx-runtime";
import { useRef } from "react";
import { ORIENTATION } from "../../../../../constants.js";
import { useOrientation } from "../../context/RangeLegendOrientationContext.js";
import { useScale } from "../../context/RangeLegendScaleContext.js";
import { useLegendFormattedLabels } from "../../hooks/use-legend-formatted-labels.js";
import { RangeLegendSegmentsItem } from "../RangeLegendSegmentsItem.js";
import { legendSvg } from "./RangeLegendSegments.sty.js";
const RangeLegendSegments = (props) => {
  const { items } = props;
  const svgRef = useRef(null);
  const { scale } = useScale();
  const { orientation } = useOrientation();
  const reversedScale = orientation === ORIENTATION.Vertical;
  const internalRangeItems = useLegendFormattedLabels(items);
  return /* @__PURE__ */ jsx(
    "svg",
    {
      ref: svgRef,
      className: legendSvg({
        isHorizontal: orientation === ORIENTATION.Horizontal
      }),
      role: "presentation",
      "data-testid": "range-legend-segments",
      children: internalRangeItems.map((item) => {
        const startPosition = scale(reversedScale ? item.to : item.from);
        const endPosition = scale(reversedScale ? item.from : item.to);
        const segmentSize = endPosition - startPosition;
        return /* @__PURE__ */ jsx(
          RangeLegendSegmentsItem,
          {
            item,
            segmentSize,
            startPosition
          },
          `segment-'${item.from}-${item.to}`
        );
      })
    }
  );
};
export {
  RangeLegendSegments
};
