// @dynatrace/strato-components/esm/charts/categorical-bar/layout/CategoricalGridZeroLineLayout.js
import { jsx } from "react/jsx-runtime";
import { plotLayout } from "../../core/components/chart-grid/ChartGrid.sty.js";
import { useGridPlotSize } from "../../core/components/chart-grid/hooks/useGridPlotSize.js";
import { GridLine } from "../../core/components/grid/GridLine.js";
import { useCategoricalBarChartBaseConfig } from "../hooks/useCategoricalBarChartBaseConfig.js";
import { useCategoricalBarChartScales } from "../hooks/useCategoricalBarChartScales.js";
import { buildZeroLine } from "../utils/build-zero-line.js";
import { isVertical } from "../utils/layout.js";
const CategoricalGridZeroLineLayout = () => {
  const { layout } = useCategoricalBarChartBaseConfig();
  const { valueScale: scale } = useCategoricalBarChartScales();
  const { width, height } = useGridPlotSize();
  const zeroLineValue = scale(0) ?? 0;
  const referenceSize = isVertical(layout) ? height : width;
  if (zeroLineValue === referenceSize || zeroLineValue === 0) {
    return null;
  }
  const {
    show: showLine,
    from,
    to
  } = buildZeroLine(layout, zeroLineValue, width, height);
  return showLine ? /* @__PURE__ */ jsx(
    "svg",
    {
      className: plotLayout,
      role: "graphics-symbol",
      "aria-label": "categorical grid zero line",
      children: /* @__PURE__ */ jsx(GridLine, { from, to, highlight: true })
    }
  ) : null;
};
CategoricalGridZeroLineLayout["displayName"] = "CategoricalGridZeroLineLayout";
export {
  CategoricalGridZeroLineLayout
};
