// @dynatrace/strato-components/charts/core/components/legend/shared/context/Legend.context.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Legend_context_exports = {};
__export(Legend_context_exports, {
  LegendContext: () => LegendContext
});
module.exports = __toCommonJS(Legend_context_exports);
var import_react = require("react");
var import_constants = require("../../../../constants.js");
const LegendContext = (0, import_react.createContext)({
  hiddenItems: /* @__PURE__ */ new Set([]),
  truncationMode: import_constants.DEFAULT_TRUNCATION_MODE,
  onLegendIconClick: () => void 0,
  onItemHover: () => void 0,
  onLegendNameClick: () => void 0,
  interactionsDisabled: false
});
LegendContext["displayName"] = "LegendContext";
