// @dynatrace/strato-components/charts/histogram/components/toolbar/HistogramToolbar.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var HistogramToolbar_exports = {};
__export(HistogramToolbar_exports, {
  HistogramToolbar: () => HistogramToolbar
});
module.exports = __toCommonJS(HistogramToolbar_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_intl = require("react-intl");
var import_DownloadDataButton = require("./DownloadDataButton.js");
var import_ExploreButton = require("./ExploreButton.js");
var import_PanButton = require("./PanButton.js");
var import_ResetButton = require("./ResetButton.js");
var import_ZoomInOutButtons = require("./ZoomInOutButtons.js");
var import_ZoomModeButton = require("./ZoomModeButton.js");
var import_useBehavioralTrackingProps = require("../../../../core/hooks/useBehavioralTrackingProps.js");
var import_ChartToolbarRenderer = require("../../../core/components/toolbar/ChartToolbarRenderer.js");
var import_get_menu_trigger_display_mode = require("../../../core/components/toolbar/utils/get-menu-trigger-display-mode.js");
var import_get_menu_trigger_icon = require("../../../core/components/toolbar/utils/get-menu-trigger-icon.js");
var import_constants = require("../../../core/constants.js");
var import_zoom = require("../../../core/utils/zoom.js");
var import_defaults = require("../../constants/defaults.js");
var import_selectors = require("../../hooks/store/selectors.js");
var import_useHistogramFilteredData = require("../../hooks/useHistogramFilteredData.js");
var import_useHistogramInteractions = require("../../hooks/useHistogramInteractions.js");
var import_useHistogramSelectionAreaRef = require("../../hooks/useHistogramSelectionAreaRef.js");
var import_useHistogramToolbar = require("../../hooks/useHistogramToolbar.js");
var import_tooltip_text = require("../../messages/tooltip-text.js");
var import_build_and_download_csv = require("../../utils/build-and-download-csv.js");
const HistogramToolbar = ({
  onMouseEnter,
  onMouseLeave,
  handlerRef
}) => {
  const toolbarContext = (0, import_useHistogramToolbar.useHistogramToolbar)();
  const { placement, collapsed, downloadData, intents } = toolbarContext;
  const interactionMode = (0, import_selectors.useInteractionModeState)();
  const { zoom, zoomX, pan } = (0, import_useHistogramInteractions.useHistogramInteractions)();
  const intl = (0, import_react_intl.useIntl)();
  const selectionAreaRef = (0, import_useHistogramSelectionAreaRef.useHistogramSelectionAreaRef)();
  const isZoomEnabled = zoom?.enabled;
  const isPanEnabled = pan?.enabled;
  const isDownloadEnabled = downloadData?.enabled ?? false;
  const anyInteractionEnabled = isZoomEnabled || zoomX?.enabled || isPanEnabled || isDownloadEnabled || intents;
  const menuIcon = (0, import_get_menu_trigger_icon.getMenuTriggerIcon)(interactionMode, isZoomEnabled);
  const menuLabel = !isPanEnabled && !isZoomEnabled ? intl.formatMessage(import_tooltip_text.HISTOGRAM_MESSAGES.zoomModeExplore) : void 0;
  const menuDisplayMode = (0, import_get_menu_trigger_display_mode.getMenuTriggerDisplayMode)({
    isPanEnabled,
    isZoomEnabled,
    isDownloadEnabled
  });
  const onClose = () => selectionAreaRef ? selectionAreaRef.current?.focus() : void 0;
  const { series } = (0, import_useHistogramFilteredData.useHistogramFilteredData)();
  const handleDownloadDataToolbar = () => (0, import_build_and_download_csv.buildAndDownloadCsv)(series);
  const showDefaultToolbarDownloadButton = isDownloadEnabled && !intents;
  const showMenuToolbarButton = isDownloadEnabled && !!intents?.length;
  const behaviorTrackingProps = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(toolbarContext)[0];
  return anyInteractionEnabled ? /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_ChartToolbarRenderer.ChartToolbarRenderer,
    {
      visible: false,
      visibleWhenMouseInBounds: true,
      handlerRef,
      placement: placement || import_defaults.defaultToolbarPlacement,
      collapsed,
      draggable: true,
      initialA11yState: (0, import_zoom.getA11yToolbarConfig)({
        zoomX,
        pan,
        zoom,
        download: downloadData,
        intent: !!intents
      }),
      zIndex: import_constants.ChartOverlaysElevation.toolbar,
      onMouseEnter,
      onMouseLeave,
      menuIcon,
      menuTriggerLabel: menuLabel,
      menuDisplayMode,
      onClose,
      isDownloadEnabled: showMenuToolbarButton,
      intents,
      handleDownloadDataToolbar,
      ...behaviorTrackingProps,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ExploreButton.ExploreButton, {}),
          isZoomEnabled && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ZoomModeButton.ZoomModeButton, {}),
          isPanEnabled && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_PanButton.PanButton, {})
        ] }),
        isZoomEnabled && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ZoomInOutButtons.ZoomInOutButtons, {}),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ResetButton.ResetButton, {}),
        showDefaultToolbarDownloadButton && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_DownloadDataButton.DownloadDataButton, {})
      ]
    }
  ) : null;
};
