// @dynatrace/strato-components/charts/histogram/components/toolbar/ZoomInOutButtons.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ZoomInOutButtons_exports = {};
__export(ZoomInOutButtons_exports, {
  ZoomInOutButtons: () => ZoomInOutButtons
});
module.exports = __toCommonJS(ZoomInOutButtons_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_intl = require("react-intl");
var import_spacings = __toESM(require("@dynatrace/strato-design-tokens/spacings"));
var import_strato_icons = require("@dynatrace/strato-icons");
var import_useGridPlotSize = require("../../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_ChartToolbarRenderer = require("../../../core/components/toolbar/ChartToolbarRenderer.js");
var import_useTriggerZoom = require("../../../core/components/zoom/hooks/useTriggerZoom.js");
var import_useA11yToolbarTabIndex = require("../../../core/hooks/useA11yToolbarTabIndex.js");
var import_zoom = require("../../../core/types/zoom.js");
var import_get_key_shortcut_by_agent = require("../../../core/utils/get-key-shortcut-by-agent.js");
var import_selectors = require("../../hooks/store/selectors.js");
var import_tooltip_text = require("../../messages/tooltip-text.js");
var import_useApplyZoomDelta = require("../overlays/hooks/useApplyZoomDelta.js");
const ZoomInOutButtons = () => {
  const tabIndexIn = (0, import_useA11yToolbarTabIndex.useA11yToolbarTabIndex)(import_zoom.ZOOM_ACTIONS.ZoomIn);
  const tabIndexOut = (0, import_useA11yToolbarTabIndex.useA11yToolbarTabIndex)(import_zoom.ZOOM_ACTIONS.ZoomOut);
  const { zoomInEnabled, zoomOutEnabled } = (0, import_selectors.useZoomState)();
  const intl = (0, import_react_intl.useIntl)();
  const { width } = (0, import_useGridPlotSize.useGridPlotSize)();
  const applyZoom = (0, import_useApplyZoomDelta.useApplyZoomDelta)();
  const triggerZoom = (0, import_useTriggerZoom.useTriggerZoom)(width, applyZoom);
  const handleZoomInClick = () => {
    triggerZoom(import_zoom.ZOOM_CHANGE_TYPE.ZoomIn);
  };
  const handleZoomInKeyDown = (event) => {
    if (event.code === "Enter") {
      triggerZoom(import_zoom.ZOOM_CHANGE_TYPE.ZoomIn);
    }
  };
  const handleZoomOutClick = () => {
    triggerZoom(import_zoom.ZOOM_CHANGE_TYPE.ZoomOut);
  };
  const handleZoomOutKeyDown = (event) => {
    if (event.code === "Enter") {
      triggerZoom(import_zoom.ZOOM_CHANGE_TYPE.ZoomOut);
    }
  };
  const zoomInShortcut = (0, import_get_key_shortcut_by_agent.getKeyShortcutByAgent)("CONTROL") + `+${intl.formatMessage(import_tooltip_text.HISTOGRAM_MESSAGES.zoomActionArrowUp)}`;
  const zoomOutShortcut = (0, import_get_key_shortcut_by_agent.getKeyShortcutByAgent)("CONTROL") + `+${intl.formatMessage(import_tooltip_text.HISTOGRAM_MESSAGES.zoomActionArrowDown)}`;
  const controlStyle = { marginRight: import_spacings.default.Size2 };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_ChartToolbarRenderer.ChartToolbarRenderer.Control,
      {
        prefixIcon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ZoomInIcon, {}),
        tooltipText: intl.formatMessage(import_tooltip_text.HISTOGRAM_MESSAGES.zoomIn),
        tooltipKeyShortcut: zoomInShortcut,
        onKeyDown: handleZoomInKeyDown,
        onClick: handleZoomInClick,
        tabIndex: tabIndexIn,
        disabled: !zoomInEnabled,
        divider: "left",
        style: controlStyle
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_ChartToolbarRenderer.ChartToolbarRenderer.Control,
      {
        prefixIcon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ZoomOutIcon, {}),
        tooltipText: intl.formatMessage(import_tooltip_text.HISTOGRAM_MESSAGES.zoomOut),
        tooltipKeyShortcut: zoomOutShortcut,
        onKeyDown: handleZoomOutKeyDown,
        onClick: handleZoomOutClick,
        tabIndex: tabIndexOut,
        disabled: !zoomOutEnabled,
        style: controlStyle
      }
    )
  ] });
};
ZoomInOutButtons["displayName"] = "ZoomInOutButtons";
