// @dynatrace/strato-components/charts/histogram/components/toolbar/ZoomModeButton.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ZoomModeButton_exports = {};
__export(ZoomModeButton_exports, {
  ZoomModeButton: () => ZoomModeButton
});
module.exports = __toCommonJS(ZoomModeButton_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_useGridPlotSize = require("../../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_types = require("../../../core/components/interactions-overlay/types.js");
var import_calculate_initial_selection_area_positions = require("../../../core/components/interactions-overlay/utils/calculate-initial-selection-area-positions.js");
var import_explore_reducers = require("../../../core/components/interactions-state-helpers/reducers/explore-reducers.js");
var import_mode_reducers = require("../../../core/components/interactions-state-helpers/reducers/mode-reducers.js");
var import_ChartToolbarRenderer = require("../../../core/components/toolbar/ChartToolbarRenderer.js");
var import_useToolbarContext = require("../../../core/components/toolbar/hooks/useToolbarContext.js");
var import_messages = require("../../../core/components/toolbar/messages.js");
var import_useA11yToolbarTabIndex = require("../../../core/hooks/useA11yToolbarTabIndex.js");
var import_get_key_shortcut_by_agent = require("../../../core/utils/get-key-shortcut-by-agent.js");
var import_get_percentage_from_pixel = require("../../../core/utils/get-percentage-from-pixel.js");
var import_selectors = require("../../hooks/store/selectors.js");
var import_state_setters = require("../../hooks/store/state-setters.js");
var import_useHistogramChartOverlayRef = require("../../hooks/useHistogramChartOverlayRef.js");
var import_tooltip_text = require("../../messages/tooltip-text.js");
const ZoomModeButton = () => {
  const setState = (0, import_state_setters.useSetState)();
  const interactionMode = (0, import_selectors.useInteractionModeState)();
  const tabIndex = (0, import_useA11yToolbarTabIndex.useA11yToolbarTabIndex)(import_types.INTERACTION_MODE.Zoom);
  const overlayRef = (0, import_useHistogramChartOverlayRef.useHistogramChartOverlayRef)();
  const { width } = (0, import_useGridPlotSize.useGridPlotSize)();
  const intl = (0, import_react_intl.useIntl)();
  const tooltipShortcut = (0, import_get_key_shortcut_by_agent.getKeyShortcutByAgent)("CONTROL") + " " + intl.formatMessage(import_messages.TOOLBAR_MESSAGES.or) + " Z";
  const { updateLastItemSelected } = (0, import_useToolbarContext.useToolbarContext)();
  const isSelected = interactionMode === import_types.INTERACTION_MODE.Zoom;
  const selectState = (0, import_selectors.useSelectState)();
  const createInitialZoomArea = () => {
    if (!(0, import_lodash_es.isNil)(overlayRef.current)) {
      const bounds = overlayRef.current.getBoundingClientRect();
      const [start, end] = (0, import_calculate_initial_selection_area_positions.calculateInitialSelectionAreaPositions)(bounds);
      setState(
        (0, import_explore_reducers.reduceExploreSelect)({
          start,
          end,
          finished: true,
          accessible: true,
          variant: "zoom",
          isKeyboardCreated: true,
          selectionPercentage: (0, import_get_percentage_from_pixel.getPercentageFromPixel)([start, end], width)
        })
      );
    }
  };
  const handleKeyDown = (event) => {
    if (event.key === "Enter") {
      createInitialZoomArea();
    }
  };
  const handleClick = () => {
    if (interactionMode !== "zoom") {
      setState((0, import_mode_reducers.reduceInteractionMode)(import_types.INTERACTION_MODE.Zoom));
    }
    if (selectState?.start && selectState?.end) {
      if (selectState?.isKeyboardCreated) {
        (0, import_mode_reducers.reduceSelectionAreaVariant)("zoom");
      } else {
        setState(import_explore_reducers.reduceExploreCancelAll);
      }
    }
    updateLastItemSelected(import_types.INTERACTION_MODE.Zoom);
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ChartToolbarRenderer.ChartToolbarRenderer.Control,
    {
      prefixIcon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ZoomToSelectionIcon, {}),
      selected: isSelected,
      tooltipText: intl.formatMessage(import_tooltip_text.HISTOGRAM_MESSAGES.zoomModeZoom),
      onClick: handleClick,
      onKeyDown: handleKeyDown,
      tabIndex,
      tooltipKeyShortcut: tooltipShortcut,
      placement: "menu"
    }
  );
};
