// @dynatrace/strato-components/tables/DataTable/features/Config.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Config_exports = {};
__export(Config_exports, {
  useDataTableConfigRef: () => useDataTableConfigRef
});
module.exports = __toCommonJS(Config_exports);
var import_react = require("react");
var import_useMergeRefs = require("../../../core/hooks/useMergeRefs.js");
var import_useDataTableConfig = require("../../hooks/useDataTableConfig.js");
var import_constants = require("../constants.js");
function useDataTableConfigRef(forwardedRef, columnSettingsModalRef, table, props) {
  const containerRef = (0, import_react.useRef)(null);
  const rootRef = (0, import_useMergeRefs.useMergeRefs)([
    forwardedRef,
    containerRef
  ]);
  const configRowThresholds = (0, import_useDataTableConfig.useDataTableConfigElement)("rowThresholds");
  (0, import_react.useImperativeHandle)(forwardedRef, () => ({
    element: containerRef.current,
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    getConfig() {
      const options = table.options;
      const state = options.state;
      return {
        pagination: {
          defaultPageIndex: state.pagination?.pageIndex,
          defaultPageSize: state.pagination?.pageSize
        },
        columnVisibility: state.columnVisibility,
        fullWidth: options.layout?.fullWidth,
        fullHeight: options.layout?.fullHeight,
        lineWrap: state.columnLineWrap,
        resizable: options.enableColumnResizing,
        selectableRows: options.enableRowSelection,
        // NOTE: typecast necessary due to generic resolving to unknown
        rowThresholds: props.rowThresholds ?? configRowThresholds,
        sortable: options.enableSorting,
        variant: state.visualVariantConfig
      };
    },
    downloadData(subset, excludeColumns) {
      switch (subset) {
        case "all":
          table.downloadAllData(excludeColumns);
          break;
        case "page":
          table.downloadPageData(excludeColumns);
          break;
        case "selected":
          table.downloadSelectedData(excludeColumns);
          break;
      }
    },
    scrollToRow: table.scrollToRow,
    getDisplayedRowIds() {
      const result = [];
      for (const row of table.getRowModel().rows) {
        if (!row[import_constants.DATA_TABLE_DETAILS_ROW_IDENTIFIER]) {
          result.push(row.id);
        }
      }
      return result;
    },
    openColumnSettings: (columnSettings) => {
      columnSettingsModalRef.current?.openColumnSettings(columnSettings);
    }
  }));
  return rootRef;
}
