// @dynatrace/strato-components/esm/charts/core/components/annotations/AnnotationsAtom.js
import { jsx, jsxs } from "react/jsx-runtime";
import { Children } from "react";
import { wrapper } from "./AnnotationsAtom.sty.js";
import { AnnotationsCanvasLayout } from "./components/canvas/AnnotationsCanvasLayout.js";
import { CanvasSvgToImagesProvider } from "./components/canvas/providers/CanvasSvgToImages.Provider.js";
import { AnnotationsOverlays } from "./components/overlays/AnnotationsOverlays.js";
import {
  DEFAULT_LABELS_WIDTH,
  TRACK_SEPARATOR_STROKE_WIDTH
} from "./constants.js";
import { AnnotationsConfigContext } from "./contexts/AnnotationsConfig.context.js";
import { MarkerCoordinatesContext } from "./contexts/MarkerCoordinates.context.js";
import { MarkerEventHandlersContext } from "./contexts/MarkerEventHandlers.context.js";
import { MarkersByTrackContext } from "./contexts/MarkersByTrack.context.js";
import { TrackEventHandlersContext } from "./contexts/TrackEventHandlers.context.js";
import { MarkerAtom } from "./slots/MarkerAtom.js";
import { TrackAtom } from "./slots/TrackAtom.js";
import { aggregateMarkersCoords } from "./utils/aggregate-markers-coords.js";
import { extractVisibleTracks } from "./utils/extract-visible-tracks.js";
import { getAnnotationsTracksHeight } from "./utils/get-annotations-tracks-height.js";
const _AnnotationsAtom = (props) => {
  const {
    width,
    children,
    showLabels = true,
    labelsWidth = DEFAULT_LABELS_WIDTH
  } = props;
  const trackWidth = width - labelsWidth;
  const tracks = extractVisibleTracks(children);
  const tracksQuantity = Children.count(tracks);
  const totalTracksHeight = getAnnotationsTracksHeight(tracksQuantity);
  const {
    markersPerTrack,
    markerEventHandlers,
    trackEventHandlers,
    icons,
    tracksConfig
  } = aggregateMarkersCoords(tracks, trackWidth);
  return /* @__PURE__ */ jsx(
    "div",
    {
      className: wrapper,
      style: {
        height: totalTracksHeight
      },
      children: /* @__PURE__ */ jsx(
        AnnotationsConfigContext.Provider,
        {
          value: {
            showLabels,
            labelsWidth,
            trackWidth
          },
          children: /* @__PURE__ */ jsx(MarkersByTrackContext.Provider, { value: tracksConfig, children: /* @__PURE__ */ jsx(MarkerCoordinatesContext.Provider, { value: markersPerTrack, children: /* @__PURE__ */ jsx(MarkerEventHandlersContext.Provider, { value: markerEventHandlers, children: /* @__PURE__ */ jsx(TrackEventHandlersContext.Provider, { value: trackEventHandlers, children: /* @__PURE__ */ jsxs(CanvasSvgToImagesProvider, { icons, children: [
            /* @__PURE__ */ jsx(
              AnnotationsOverlays,
              {
                height: totalTracksHeight - TRACK_SEPARATOR_STROKE_WIDTH
              }
            ),
            /* @__PURE__ */ jsx(
              AnnotationsCanvasLayout,
              {
                width,
                height: totalTracksHeight
              }
            )
          ] }) }) }) }) })
        }
      )
    }
  );
};
const AnnotationsAtom = Object.assign(_AnnotationsAtom, {
  TrackAtom,
  MarkerAtom
});
AnnotationsAtom.displayName = "AnnotationsAtom";
export {
  AnnotationsAtom,
  _AnnotationsAtom
};
