// @dynatrace/strato-components/esm/charts/core/components/annotations/constants.js
import Colors from "@dynatrace/strato-design-tokens/colors";
const DEFAULT_LABELS_WIDTH = 60;
const DEFAULT_LABELS_OFFSET = 20;
const TRACK_HEIGHT = 24;
const TRACK_SEPARATOR_COLOR = Colors.Border.Neutral.Default;
const EMPHASIZED_TRACK_SEPARATOR_COLOR = Colors.Border.Neutral.Accent;
const TRACK_SEPARATOR_STROKE_WIDTH = 1;
const TRACK_PADDING_Y = 1;
const LABEL_OFFSET = 12;
const PILL_MARKER_HEIGHT = 20;
const OUTLINE_STROKE_WIDTH = 2;
const DEFAULT_MARKER_COLOR = Colors.Charts.Categorical.Color01.Default;
const MARKER_RADIUS = 10;
const MINIMUM_MARKER_WIDTH = 3;
const MARKER_OPACITY = 0.85;
const MARKER_STROKE_WIDTH = 1;
const MARKER_STROKE_COLOR = Colors.Background.Base.Default;
const OUTER_CIRCLE_RADIUS = MARKER_RADIUS + OUTLINE_STROKE_WIDTH;
const SCROLLBAR_THRESHOLD = 15;
const SCROLLBAR_WIDTH = 8;
const DEFAULT_ICON_COLOR = Colors.Text.Neutral.OnAccent.Default;
const MARKER_PROPS = {
  opacity: MARKER_OPACITY,
  stroke: MARKER_STROKE_COLOR,
  strokeWidth: MARKER_STROKE_WIDTH
};
const DEFAULT_TRACK_HIDDEN = false;
const DEFAULT_MARKER_PRIORITY = 0;
const DEFAULT_MARKER_HIDDEN = false;
const DEFAULT_TOOLTIP_Y_OFFSET = 40;
const DEFAULT_SHOW_LABELS = true;
const DEFAULT_TEXT_OVERFLOW = "truncate";
const DEFAULT_TRUNCATION_MODE = "middle";
const DEFAULT_TOOLTIP_MAX_LINES = 1;
const DEFAULT_INTEGRATION_TEXT_OVERFLOW = "expand";
var TestIds = /* @__PURE__ */ ((TestIds2) => {
  TestIds2["Marker"] = "marker";
  TestIds2["Track"] = "track";
  TestIds2["MarkerGroup"] = "marker-group";
  TestIds2["AnnotationIndicator"] = "annotation-indicator";
  return TestIds2;
})(TestIds || {});
const MARKER_ICON_OFFSET = 2;
const MARKER_LABEL_PADDING = 6;
const FOREIGN_OBJECT_THRESHOLD = 8;
const MARKER_BORDER_RADIUS = 2;
const MARKER_CONTENT_TO_BORDER_PADDING = 4;
const MARGIN_TOP_STROKE = 1;
export {
  DEFAULT_ICON_COLOR,
  DEFAULT_INTEGRATION_TEXT_OVERFLOW,
  DEFAULT_LABELS_OFFSET,
  DEFAULT_LABELS_WIDTH,
  DEFAULT_MARKER_COLOR,
  DEFAULT_MARKER_HIDDEN,
  DEFAULT_MARKER_PRIORITY,
  DEFAULT_SHOW_LABELS,
  DEFAULT_TEXT_OVERFLOW,
  DEFAULT_TOOLTIP_MAX_LINES,
  DEFAULT_TOOLTIP_Y_OFFSET,
  DEFAULT_TRACK_HIDDEN,
  DEFAULT_TRUNCATION_MODE,
  EMPHASIZED_TRACK_SEPARATOR_COLOR,
  FOREIGN_OBJECT_THRESHOLD,
  LABEL_OFFSET,
  MARGIN_TOP_STROKE,
  MARKER_BORDER_RADIUS,
  MARKER_CONTENT_TO_BORDER_PADDING,
  MARKER_ICON_OFFSET,
  MARKER_LABEL_PADDING,
  MARKER_OPACITY,
  MARKER_PROPS,
  MARKER_RADIUS,
  MARKER_STROKE_COLOR,
  MARKER_STROKE_WIDTH,
  MINIMUM_MARKER_WIDTH,
  OUTER_CIRCLE_RADIUS,
  OUTLINE_STROKE_WIDTH,
  PILL_MARKER_HEIGHT,
  SCROLLBAR_THRESHOLD,
  SCROLLBAR_WIDTH,
  TRACK_HEIGHT,
  TRACK_PADDING_Y,
  TRACK_SEPARATOR_COLOR,
  TRACK_SEPARATOR_STROKE_WIDTH,
  TestIds
};
