// @dynatrace/strato-components/charts/xy-chart/components/tooltip/HoveredPointTooltip/hoveredPointTooltipBodyUtils.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var hoveredPointTooltipBodyUtils_exports = {};
__export(hoveredPointTooltipBodyUtils_exports, {
  buildHoveredPointTooltipBodyProps: () => buildHoveredPointTooltipBodyProps
});
module.exports = __toCommonJS(hoveredPointTooltipBodyUtils_exports);
var import_safe_get = require("../../../../core/utils/safe-get.js");
var import_guards = require("../../../types/guards.js");
var import_value_representation = require("../../../utils/value-representation.js");
var import_get_tooltip_formatter = require("../utils/get-tooltip-formatter.js");
var import_getHoveredPointTooltipBodyContent = require("./hooks/getHoveredPointTooltipBodyContent.js");
const buildHoveredPointTooltipBodyProps = ({
  datapoint,
  series,
  seriesToMetadata,
  axisIdToFormatter,
  axisIdToScaleDomain,
  datapointMetadataMap,
  formatCategory,
  intl
}) => {
  const { xAxis, yAxis } = (0, import_safe_get.safeGet)(seriesToMetadata, series);
  const xFormatter = (0, import_get_tooltip_formatter.getTooltipFormatter)(xAxis, axisIdToFormatter, intl);
  const yFormatter = (0, import_get_tooltip_formatter.getTooltipFormatter)(yAxis, axisIdToFormatter, intl);
  const datapointMetadata = datapointMetadataMap.get(datapoint);
  const binDomain = datapointMetadata?.binDomain;
  const isRelative = typeof binDomain === "number" && binDomain;
  const yDomain = isRelative && datapointMetadata?.baseline === "y" ? {
    min: binDomain * -1,
    max: binDomain
  } : (0, import_safe_get.safeGet)(axisIdToScaleDomain, xAxis.id);
  const xDomain = isRelative && datapointMetadata?.baseline === "x" ? {
    min: binDomain * -1,
    max: binDomain
  } : (0, import_safe_get.safeGet)(axisIdToScaleDomain, yAxis.id);
  const { x0, x1, y0, y1 } = datapoint;
  const buildCategoricalStringFormatter = (axis) => (0, import_guards.isCategorical)(axis.type) ? (value) => formatCategory(axis, value) ?? value : void 0;
  const xResolvedFormatter = buildCategoricalStringFormatter(xAxis) ?? xFormatter;
  const yResolvedFormatter = buildCategoricalStringFormatter(yAxis) ?? yFormatter;
  const { text: textX, duration: durationX } = (0, import_getHoveredPointTooltipBodyContent.getHoveredPointTooltipBodyContent)(
    intl,
    x0,
    x1,
    xResolvedFormatter,
    (0, import_value_representation.getRelativeAccessor)(xAxis, yDomain)
  );
  const { text: textY, duration: durationY } = (0, import_getHoveredPointTooltipBodyContent.getHoveredPointTooltipBodyContent)(
    intl,
    y0,
    y1,
    yResolvedFormatter,
    (0, import_value_representation.getRelativeAccessor)(yAxis, xDomain)
  );
  return {
    xAxisLabel: xAxis.label,
    yAxisLabel: yAxis.label,
    xAxisValue: textX,
    yAxisValue: textY,
    xAxisDuration: durationX,
    yAxisDuration: durationY
  };
};
