// @dynatrace/strato-components/charts/xy-chart/components/tooltip/HoveredPointTooltip/hoveredPointTooltipHeaderUtils.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var hoveredPointTooltipHeaderUtils_exports = {};
__export(hoveredPointTooltipHeaderUtils_exports, {
  buildHoveredPointTooltipHeaderProps: () => buildHoveredPointTooltipHeaderProps
});
module.exports = __toCommonJS(hoveredPointTooltipHeaderUtils_exports);
var import_useDefaultAction = require("../../../../core/components/series-action/hooks/useDefaultAction.js");
var import_get_formatter = require("../../../../core/utils/get-formatter.js");
var import_safe_get = require("../../../../core/utils/safe-get.js");
var import_build_series_action_item = require("../../../utils/build-series-action-item.js");
var import_get_color_from_scale_and_data = require("../../../utils/coloring/get-color-from-scale-and-data.js");
var import_type_guards = require("../../../utils/type-guards.js");
const buildHoveredPointTooltipHeaderProps = ({
  datapoint,
  series,
  seriesToMetadata,
  seriesWithDatapoints,
  datapointMetadataMap,
  colorScaleMap,
  truncationMode,
  idFactory = () => crypto.randomUUID()
}) => {
  let content;
  let valueToDisplay = "";
  let colorToDisplay = "";
  let actions;
  const metadata = (0, import_safe_get.safeGet)(seriesToMetadata, series);
  const enrichedDatapoint = {
    series,
    datapoint,
    metadata
  };
  if ((0, import_type_guards.isRectTooltipContext)(enrichedDatapoint)) {
    const colorScale = (0, import_safe_get.safeGet)(colorScaleMap, series.id);
    const datapointMetadata = (0, import_safe_get.safeGet)(datapointMetadataMap, datapoint);
    const rawValue = enrichedDatapoint.datapoint.value;
    const { valueAccessorLabel, valueAccessor, valueFormatter, valueUnit } = enrichedDatapoint.metadata;
    const tooltipName = valueAccessorLabel ?? valueAccessor ?? "";
    content = {
      id: idFactory(),
      name: tooltipName
    };
    valueToDisplay = typeof rawValue === "number" ? (0, import_get_formatter.getFormatter)(valueFormatter)(rawValue, valueUnit) : rawValue?.toString() ?? "";
    colorToDisplay = rawValue !== void 0 ? (0, import_get_color_from_scale_and_data.getColorFromScaleAndData)(
      colorScale,
      enrichedDatapoint.datapoint.value,
      datapointMetadata.colorOverride
    ) ?? "" : "";
    const metadataActions = enrichedDatapoint.metadata.actions;
    actions = metadataActions ? () => metadataActions(enrichedDatapoint.datapoint) : void 0;
  } else if ((0, import_type_guards.isBarTooltipContext)(enrichedDatapoint)) {
    const { color, name } = metadata;
    content = {
      id: idFactory(),
      name: name ?? ""
    };
    colorToDisplay = color?.Default ?? "";
    const metadataActions = enrichedDatapoint.metadata.actions;
    actions = metadataActions ? () => {
      const actionItem = (0, import_build_series_action_item.buildBarSeriesActionItem)({
        datapoint: enrichedDatapoint.datapoint,
        allSeries: seriesWithDatapoints,
        seriesToMetadata,
        datapointMetadataMap,
        currentSeries: enrichedDatapoint.series
      });
      return metadataActions(actionItem);
    } : void 0;
  } else {
    content = {
      id: idFactory(),
      name: ""
    };
  }
  return {
    color: colorToDisplay,
    name: content.name,
    value: valueToDisplay,
    actions,
    defaultActions: (0, import_useDefaultAction.getDefaultAction)(content),
    truncationMode
  };
};
