// @dynatrace/strato-components/charts/xy-chart/components/tooltip/HoveredPointTooltip/HoveredPointTooltip.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var HoveredPointTooltip_exports = {};
__export(HoveredPointTooltip_exports, {
  HoveredPointTooltip: () => HoveredPointTooltip
});
module.exports = __toCommonJS(HoveredPointTooltip_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_intl = require("react-intl");
var import_hoveredPointTooltipBodyUtils = require("./hoveredPointTooltipBodyUtils.js");
var import_hoveredPointTooltipHeaderUtils = require("./hoveredPointTooltipHeaderUtils.js");
var import_OverlayTooltip = require("../../../../core/components/overlay-tooltip/OverlayTooltip.js");
var import_useAxesFormatter = require("../../../hooks/useAxesFormatter.js");
var import_useBarData = require("../../../hooks/useBarData.js");
var import_useFormatCategory = require("../../../hooks/useFormatCategory.js");
var import_useXYChartColorScales = require("../../../hooks/useXYChartColorScales.js");
var import_useXYChartData = require("../../../hooks/useXYChartData.js");
var import_useXYChartTruncationMode = require("../../../hooks/useXYChartTruncationMode.js");
function HoveredPointTooltip({
  datapoint,
  series,
  seriesToMetadata
}) {
  const { filteredSeriesWithDatapoints } = (0, import_useBarData.useBarData)();
  const { axisIdToScaleDomain, datapointMetadataMap } = (0, import_useXYChartData.useXYChartData)();
  const colorScaleMap = (0, import_useXYChartColorScales.useXYChartColorScales)();
  const { truncationMode } = (0, import_useXYChartTruncationMode.useXYChartTruncationMode)();
  const axisIdToFormatter = (0, import_useAxesFormatter.useAxesFormatter)();
  const formatCategory = (0, import_useFormatCategory.useFormatCategory)();
  const intl = (0, import_react_intl.useIntl)();
  const { name, value, color, actions } = (0, import_hoveredPointTooltipHeaderUtils.buildHoveredPointTooltipHeaderProps)({
    datapoint,
    series,
    seriesToMetadata,
    seriesWithDatapoints: filteredSeriesWithDatapoints,
    datapointMetadataMap,
    colorScaleMap
  });
  const { yAxisValue, xAxisValue, xAxisDuration, yAxisDuration } = (0, import_hoveredPointTooltipBodyUtils.buildHoveredPointTooltipBodyProps)({
    datapoint,
    series,
    seriesToMetadata,
    axisIdToFormatter,
    axisIdToScaleDomain,
    datapointMetadataMap,
    formatCategory,
    intl
  });
  const isRect = series.type === "rect";
  const yText = `${yAxisValue} ${yAxisDuration ? `(${yAxisDuration})` : ""}`;
  const data = [
    {
      name: isRect ? yText : name,
      value: isRect ? value : yText,
      color
    }
  ];
  const label = datapoint.x0 instanceof Date ? "" : void 0;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_OverlayTooltip.OverlayTooltip, { truncation: truncationMode, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_OverlayTooltip.OverlayTooltip.Header,
      {
        header1: xAxisValue,
        duration1: xAxisDuration,
        label
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_OverlayTooltip.OverlayTooltip.Series,
      {
        data,
        variant: "shared",
        activeItem: -1,
        seriesItemTemplate: (item) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayTooltip.OverlayTooltip.SeriesItem, { ...item, actions })
      }
    )
  ] });
}
