// @dynatrace/strato-geo/map/components/ChoroplethLayer/ChoroplethLayer.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ChoroplethLayer_exports = {};
__export(ChoroplethLayer_exports, {
  ChoroplethLayer: () => ChoroplethLayer
});
module.exports = __toCommonJS(ChoroplethLayer_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_maplibre = require("@vis.gl/react-maplibre");
var import_core = require("@dynatrace/strato-components/core");
var import_ChoroplethLayerTooltip = require("./ChoroplethLayerTooltip.js");
var import_ChoroplethOutlineLayer = require("./ChoroplethOutlineLayer.js");
var import_parse_region_data_to_geo_json = require("./utils/parse-region-data-to-geo-json.js");
var import_constants = require("../../constants.js");
var import_use_layer_coloring_strategy = require("../../hooks/use-layer-coloring-strategy.js");
var import_use_legend_interaction_highlighting = require("../../hooks/use-legend-interaction-highlighting.js");
var import_use_legend_interaction_visibility = require("../../hooks/use-legend-interaction-visibility.js");
var import_use_map_base_layer_features = require("../../hooks/use-map-base-layer-features.js");
var import_use_resolve_color = require("../../hooks/use-resolve-color.js");
var import_use_tooltip_template = require("../../hooks/use-tooltip-template.js");
var import_ChoroplethLayerTooltip2 = require("../../slots/ChoroplethLayerTooltip.js");
var import_tooltip_type_guards = require("../../utils/tooltip-type-guards.js");
var import_TooltipWrapper = require("../TooltipWrapper.js");
const ChoroplethLayer = (props) => {
  const { layerId, data, regionAccessor, children } = props;
  const [behavioralTrackingProps] = (0, import_core._useBehavioralTrackingProps)(props);
  const mapSource = (0, import_use_map_base_layer_features.useMapBaseLayerFeatures)();
  const colorParser = (0, import_use_layer_coloring_strategy.useLayerColoringStrategy)();
  const choroplethColor = (0, import_use_resolve_color.useResolveChoroplethColor)(
    import_constants.DEFAULT_CHOROPLETH_COLOR,
    props,
    colorParser
  );
  const sourceId = layerId.replace("layer", "source");
  (0, import_use_legend_interaction_highlighting.useLegendInteractionHighlighting)(props, sourceId);
  (0, import_use_legend_interaction_visibility.useLegendInteractionVisibility)(props, layerId);
  const dataGeoJSON = mapSource && (0, import_parse_region_data_to_geo_json.parseRegionDataToGeoJSON)(mapSource, data, regionAccessor, choroplethColor);
  const {
    handler: tooltipTemplate,
    hidden: tooltipHidden,
    symbolAlignment
  } = (0, import_use_tooltip_template.useTooltipTemplate)(children, import_ChoroplethLayerTooltip2.ChoroplethLayerTooltip, {
    extractProps: true
  });
  return dataGeoJSON && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_react_maplibre.Source, { id: sourceId, type: "geojson", data: dataGeoJSON, generateId: true, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_react_maplibre.Layer,
      {
        id: layerId,
        source: sourceId,
        type: "fill",
        paint: {
          "fill-color": ["get", "__color"],
          "fill-opacity": [
            "case",
            ["==", ["feature-state", "legendHover"], false],
            import_constants.SHAPE_OPACITY_DIMMED,
            ["==", ["feature-state", "active"], true],
            import_constants.SHAPE_OPACITY_DEFAULT,
            ["==", ["feature-state", "hover"], true],
            import_constants.SHAPE_OPACITY_DEFAULT,
            ["==", ["feature-state", "legendHover"], true],
            import_constants.SHAPE_OPACITY_DEFAULT,
            ["==", ["feature-state", "isAnyActive"], true],
            import_constants.SHAPE_OPACITY_DIMMED,
            import_constants.SHAPE_OPACITY_DEFAULT
          ]
        },
        ...behavioralTrackingProps
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ChoroplethOutlineLayer.ChoroplethOutlineLayer, { sourceId }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TooltipWrapper.TooltipWrapper, { layerId, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_ChoroplethLayerTooltip.ChoroplethLayerTooltip,
      {
        tooltipTemplate: (0, import_tooltip_type_guards.isChoroplethTooltipTemplate)(tooltipTemplate) ? tooltipTemplate : void 0,
        hidden: tooltipHidden,
        symbolAlignment
      }
    ) })
  ] });
};
ChoroplethLayer["displayName"] = "ChoroplethLayer";
