// @dynatrace/strato-components/layouts/help-menu/HelpMenu.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var HelpMenu_exports = {};
__export(HelpMenu_exports, {
  HelpMenu: () => HelpMenu
});
module.exports = __toCommonJS(HelpMenu_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_app_environment = require("@dynatrace-sdk/app-environment");
var import_navigation = require("@dynatrace-sdk/navigation");
var import_Button = require("../../buttons/button/Button.js");
var import_Menu = require("../../navigation/menu/Menu.js");
var import_Tooltip = require("../../overlays/tooltip/Tooltip.js");
var import_Flex = require("../flex/Flex.js");
const COMPONENT_NAME = "HelpMenu";
const MENU_ITEMS = [
  {
    label: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_intl.FormattedMessage, { defaultMessage: "What's new", id: "TY3F/kE6n91t0TL0" }),
    icon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.WhatsNewIcon, {}),
    key: "whatsNew"
  },
  {
    label: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_intl.FormattedMessage, { defaultMessage: "Get started", id: "kOQNUENAg02eeaTY" }),
    icon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.GuideIcon, {}),
    key: "getStarted"
  },
  {
    label: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_intl.FormattedMessage, { defaultMessage: "Documentation", id: "npzzIhowJGIZhj8d" }),
    icon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.DescriptionIcon, {}),
    key: "documentation"
  },
  {
    label: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_react_intl.FormattedMessage,
      {
        defaultMessage: "Keyboard shortcuts",
        id: "JmaVR5qUnuxE3qnb"
      }
    ),
    icon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.KeyboardIcon, {}),
    key: "keyboardShortcuts"
  },
  {
    label: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_react_intl.FormattedMessage,
      {
        defaultMessage: "Try in Playground",
        id: "Nar807M4NJhqMw/T"
      }
    ),
    icon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.PlayIcon, {}),
    key: "playground"
  },
  {
    label: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_intl.FormattedMessage, { defaultMessage: "Share feedback", id: "Ia87HCOLBoxMcV9t" }),
    icon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.FeedbackIcon, {}),
    key: "feedback"
  },
  {
    label: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_intl.FormattedMessage, { defaultMessage: "About this app", id: "9/nhbjHpPPAhMbHo" }),
    icon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.InformationIcon, {}),
    key: "about"
  }
];
const HelpMenuFooter = () => {
  const version = (0, import_app_environment.getAppVersion)();
  if (!version) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Label, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      style: {
        maxWidth: 500,
        overflow: "hidden",
        textOverflow: "ellipsis",
        whiteSpace: "nowrap"
      },
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_react_intl.FormattedMessage,
        {
          defaultMessage: "Version {version}",
          description: "App version label",
          id: "W2KEm8AIcM8pOixf",
          values: { version }
        }
      )
    }
  ) });
};
const HelpMenuLink = ({
  href,
  target = "_blank",
  onSelect,
  label,
  icon,
  ...rest
}) => {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_Menu.Menu.Link,
    {
      ...rest,
      onClick: () => onSelect?.(),
      href,
      target,
      children: [
        icon ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Prefix, { children: icon }) : null,
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { justifyContent: "space-between", width: "inherit", children: [
          label,
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ExternalLinkIcon, { style: { flexShrink: 0 } })
        ] })
      ]
    }
  );
};
const HelpMenuButton = ({
  onSelect,
  label,
  icon,
  ...rest
}) => {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Menu.Menu.Item, { onSelect, ...rest, children: [
    icon ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Prefix, { children: icon }) : null,
    label
  ] });
};
const HelpMenu = (props) => {
  const {
    entries,
    onOpenChange,
    children,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow
  } = props;
  const [isOpen, setIsOpen] = (0, import_react.useState)(false);
  const intl = (0, import_react_intl.useIntl)();
  const label = intl.formatMessage({
    defaultMessage: "Help",
    id: "xHrhU3DP4e0nge7c"
  });
  const handleOpenChange = (value) => {
    setIsOpen(value);
    onOpenChange?.(value);
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_Menu.Menu,
    {
      open: isOpen,
      onOpenChange: handleOpenChange,
      "data-dt-component": COMPONENT_NAME,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Tooltip.Tooltip, { text: label, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_Menu.Menu.Trigger,
          {
            "data-dt-component": COMPONENT_NAME,
            "data-dt-name": "OpenHelpMenu",
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button, { "aria-label": label, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.HelpIcon, {}) }) })
          }
        ) }),
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Menu.Menu.Content, { children: [
          MENU_ITEMS.map(({ label: label2, key, icon }) => {
            const entry = entries[key];
            if (!entry) {
              return null;
            }
            if (entry === "default") {
              return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                HelpMenuDefaults,
                {
                  entryKey: key,
                  icon,
                  label: label2
                },
                key
              );
            }
            if (Array.isArray(entry)) {
              return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Menu.Menu.Sub, { children: [
                /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Menu.Menu.SubTrigger, { children: [
                  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Prefix, { children: icon }),
                  label2
                ] }),
                /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.SubContent, { children: entry.map((entryChild) => {
                  if ("href" in entryChild) {
                    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                      HelpMenuLink,
                      {
                        label: entryChild.label,
                        onSelect: entryChild.onSelect,
                        href: entryChild.href,
                        target: entryChild.target
                      },
                      entryChild.label
                    );
                  }
                  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                    HelpMenuButton,
                    {
                      label: entryChild.label,
                      onSelect: entryChild.onSelect
                    },
                    entryChild.label
                  );
                }) })
              ] }, key);
            }
            if ("href" in entry) {
              return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                HelpMenuLink,
                {
                  label: label2,
                  icon,
                  onSelect: entry.onSelect,
                  href: entry.href,
                  target: entry.target
                },
                key
              );
            }
            return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              HelpMenuButton,
              {
                label: label2,
                icon,
                onSelect: entry.onSelect
              },
              key
            );
          }),
          children,
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(HelpMenuFooter, {})
        ] })
      ]
    }
  );
};
const HelpMenuDefaults = (props) => {
  switch (props.entryKey) {
    case "whatsNew":
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        HelpMenuButton,
        {
          onSelect: () => (0, import_navigation.sendIntent)(
            {
              appId: (0, import_app_environment.getAppId)(),
              version: (0, import_app_environment.getAppVersion)()
            },
            {
              recommendedAppId: "dynatrace.hub",
              recommendedIntentId: "view_app_releases_add_on_intent"
            }
          ),
          label: props.label,
          icon: props.icon
        }
      );
    case "keyboardShortcuts":
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        HelpMenuButton,
        {
          onSelect: () => {
            globalThis.dtRuntime?.uiCommands?.showUiCommands?.();
          },
          label: props.label,
          icon: props.icon
        }
      );
    case "playground":
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        HelpMenuLink,
        {
          href: `https://wkf10640.apps.dynatrace.com/ui/apps/${(0, import_app_environment.getAppId)()}`,
          label: props.label,
          icon: props.icon
        }
      );
    case "about":
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        HelpMenuButton,
        {
          onSelect: () => {
            (0, import_navigation.sendIntent)(
              { appId: (0, import_app_environment.getAppId)() },
              {
                recommendedAppId: "dynatrace.hub",
                recommendedIntentId: "view_app_details"
              }
            );
          },
          label: props.label,
          icon: props.icon
        }
      );
    default:
      return null;
  }
};
