// @dynatrace/strato-components/esm/tables/DataTable/tanstack-table-core-override/src/core/column.js
import { getMemoOptions, memo } from "../utils";
function createColumn(table, columnDef, depth, parent) {
  const defaultColumn = table._getDefaultColumnDef();
  const resolvedColumnDef = {
    ...defaultColumn,
    ...columnDef
  };
  const accessorKey = resolvedColumnDef.accessorKey;
  let id = resolvedColumnDef.id ?? (accessorKey ? typeof String.prototype.replaceAll === "function" ? accessorKey.replaceAll(".", "_") : accessorKey.replace(/\./g, "_") : void 0) ?? (typeof resolvedColumnDef.header === "string" ? resolvedColumnDef.header : void 0);
  let accessorFn;
  if (resolvedColumnDef.accessorFn) {
    accessorFn = resolvedColumnDef.accessorFn;
  } else if (accessorKey) {
    if (accessorKey.includes(".")) {
      accessorFn = (originalRow) => {
        let result = originalRow;
        for (const key of accessorKey.split(".")) {
          result = result?.[key];
          if (result === void 0) {
            console.warn(
              `"${key}" in deeply nested key "${accessorKey}" returned undefined.`
            );
          }
        }
        return result;
      };
    } else {
      accessorFn = (originalRow) => originalRow[resolvedColumnDef.accessorKey];
    }
  }
  if (!id) {
    if (true) {
      throw new Error(
        resolvedColumnDef.accessorFn ? `Columns require an id when using an accessorFn` : `Columns require an id when using a non-string header`
      );
    }
    throw new Error();
  }
  let column = {
    id: `${String(id)}`,
    accessorFn,
    parent,
    depth,
    columnDef: resolvedColumnDef,
    columns: [],
    getFlatColumns: memo(
      () => [true],
      () => {
        return [
          column,
          ...column.columns?.flatMap((d) => d.getFlatColumns())
        ];
      },
      getMemoOptions(table.options, "debugColumns", "column.getFlatColumns")
    ),
    getLeafColumns: memo(
      () => [table._getOrderColumnsFn()],
      (orderColumns) => {
        if (column.columns?.length) {
          let leafColumns = column.columns.flatMap(
            (column2) => column2.getLeafColumns()
          );
          return orderColumns(leafColumns);
        }
        return [column];
      },
      getMemoOptions(table.options, "debugColumns", "column.getLeafColumns")
    )
  };
  for (const feature of table._features) {
    feature.createColumn?.(column, table);
  }
  return column;
}
export {
  createColumn
};
