// @dynatrace/strato-components/esm/tables/DataTable/tanstack-table-core-override/src/core/headers.js
import { getMemoOptions, memo } from "../utils";
const debug = "debugHeaders";
function createHeader(table, column, options) {
  const id = options.id ?? column.id;
  let header = {
    id,
    column,
    index: options.index,
    isPlaceholder: !!options.isPlaceholder,
    placeholderId: options.placeholderId,
    depth: options.depth,
    subHeaders: [],
    colSpan: 0,
    rowSpan: 0,
    headerGroup: null,
    getLeafHeaders: () => {
      const leafHeaders = [];
      const recurseHeader = (h) => {
        if (h.subHeaders && h.subHeaders.length) {
          h.subHeaders.map(recurseHeader);
        }
        leafHeaders.push(h);
      };
      recurseHeader(header);
      return leafHeaders;
    },
    getContext: () => ({
      table,
      header,
      column
    })
  };
  table._features.forEach((feature) => {
    feature.createHeader?.(header, table);
  });
  return header;
}
const Headers = {
  createTable: (table) => {
    table.getHeaderGroups = memo(
      () => [
        table.getAllColumns(),
        table.getVisibleLeafColumns(),
        table.getState().columnPinning.left,
        table.getState().columnPinning.right
      ],
      (allColumns, leafColumns, left, right) => {
        const leftColumns = left?.map((columnId) => leafColumns.find((d) => d.id === columnId)).filter(Boolean) ?? [];
        const rightColumns = right?.map((columnId) => leafColumns.find((d) => d.id === columnId)).filter(Boolean) ?? [];
        const centerColumns = leafColumns.filter(
          (column) => !left?.includes(column.id) && !right?.includes(column.id)
        );
        const headerGroups = buildHeaderGroups(
          allColumns,
          [...leftColumns, ...centerColumns, ...rightColumns],
          table
        );
        return headerGroups;
      },
      getMemoOptions(table.options, debug, "getHeaderGroups")
    );
    table.getCenterHeaderGroups = memo(
      () => [
        table.getAllColumns(),
        table.getVisibleLeafColumns(),
        table.getState().columnPinning.left,
        table.getState().columnPinning.right
      ],
      (allColumns, leafColumns, left, right) => {
        leafColumns = leafColumns.filter(
          (column) => !left?.includes(column.id) && !right?.includes(column.id)
        );
        return buildHeaderGroups(allColumns, leafColumns, table, "center");
      },
      getMemoOptions(table.options, debug, "getCenterHeaderGroups")
    );
    table.getLeftHeaderGroups = memo(
      () => [
        table.getAllColumns(),
        table.getVisibleLeafColumns(),
        table.getState().columnPinning.left
      ],
      (allColumns, leafColumns, left) => {
        const orderedLeafColumns = left?.map((columnId) => leafColumns.find((d) => d.id === columnId)).filter(Boolean) ?? [];
        return buildHeaderGroups(allColumns, orderedLeafColumns, table, "left");
      },
      getMemoOptions(table.options, debug, "getLeftHeaderGroups")
    );
    table.getRightHeaderGroups = memo(
      () => [
        table.getAllColumns(),
        table.getVisibleLeafColumns(),
        table.getState().columnPinning.right
      ],
      (allColumns, leafColumns, right) => {
        const orderedLeafColumns = right?.map((columnId) => leafColumns.find((d) => d.id === columnId)).filter(Boolean) ?? [];
        return buildHeaderGroups(
          allColumns,
          orderedLeafColumns,
          table,
          "right"
        );
      },
      getMemoOptions(table.options, debug, "getRightHeaderGroups")
    );
    table.getFooterGroups = memo(
      () => [table.getHeaderGroups()],
      (headerGroups) => {
        return [...headerGroups].reverse();
      },
      getMemoOptions(table.options, debug, "getFooterGroups")
    );
    table.getLeftFooterGroups = memo(
      () => [table.getLeftHeaderGroups()],
      (headerGroups) => {
        return [...headerGroups].reverse();
      },
      getMemoOptions(table.options, debug, "getLeftFooterGroups")
    );
    table.getCenterFooterGroups = memo(
      () => [table.getCenterHeaderGroups()],
      (headerGroups) => {
        return [...headerGroups].reverse();
      },
      getMemoOptions(table.options, debug, "getCenterFooterGroups")
    );
    table.getRightFooterGroups = memo(
      () => [table.getRightHeaderGroups()],
      (headerGroups) => {
        return [...headerGroups].reverse();
      },
      getMemoOptions(table.options, debug, "getRightFooterGroups")
    );
    table.getFlatHeaders = memo(
      () => [table.getHeaderGroups()],
      (headerGroups) => {
        return headerGroups.map((headerGroup) => {
          return headerGroup.headers;
        }).flat();
      },
      getMemoOptions(table.options, debug, "getFlatHeaders")
    );
    table.getLeftFlatHeaders = memo(
      () => [table.getLeftHeaderGroups()],
      (left) => {
        return left.map((headerGroup) => {
          return headerGroup.headers;
        }).flat();
      },
      getMemoOptions(table.options, debug, "getLeftFlatHeaders")
    );
    table.getCenterFlatHeaders = memo(
      () => [table.getCenterHeaderGroups()],
      (left) => {
        return left.map((headerGroup) => {
          return headerGroup.headers;
        }).flat();
      },
      getMemoOptions(table.options, debug, "getCenterFlatHeaders")
    );
    table.getRightFlatHeaders = memo(
      () => [table.getRightHeaderGroups()],
      (left) => {
        return left.map((headerGroup) => {
          return headerGroup.headers;
        }).flat();
      },
      getMemoOptions(table.options, debug, "getRightFlatHeaders")
    );
    table.getCenterLeafHeaders = memo(
      () => [table.getCenterFlatHeaders()],
      (flatHeaders) => {
        return flatHeaders.filter((header) => !header.subHeaders?.length);
      },
      getMemoOptions(table.options, debug, "getCenterLeafHeaders")
    );
    table.getLeftLeafHeaders = memo(
      () => [table.getLeftFlatHeaders()],
      (flatHeaders) => {
        return flatHeaders.filter((header) => !header.subHeaders?.length);
      },
      getMemoOptions(table.options, debug, "getLeftLeafHeaders")
    );
    table.getRightLeafHeaders = memo(
      () => [table.getRightFlatHeaders()],
      (flatHeaders) => {
        return flatHeaders.filter((header) => !header.subHeaders?.length);
      },
      getMemoOptions(table.options, debug, "getRightLeafHeaders")
    );
    table.getLeafHeaders = memo(
      () => [
        table.getLeftHeaderGroups(),
        table.getCenterHeaderGroups(),
        table.getRightHeaderGroups()
      ],
      (left, center, right) => {
        return [
          ...left[0]?.headers ?? [],
          ...center[0]?.headers ?? [],
          ...right[0]?.headers ?? []
        ].map((header) => {
          return header.getLeafHeaders();
        }).flat();
      },
      getMemoOptions(table.options, debug, "getLeafHeaders")
    );
  }
};
function buildHeaderGroups(allColumns, columnsToGroup, table, headerFamily) {
  let maxDepth = 0;
  const findMaxDepth = (columns, depth = 1) => {
    maxDepth = Math.max(maxDepth, depth);
    columns.filter((column) => column.getIsVisible()).forEach((column) => {
      if (column.columns?.length) {
        findMaxDepth(column.columns, depth + 1);
      }
    }, 0);
  };
  findMaxDepth(allColumns);
  let headerGroups = [];
  const createHeaderGroup = (headersToGroup, depth) => {
    const headerGroup = {
      depth,
      id: [headerFamily, `${depth}`].filter(Boolean).join("_"),
      headers: []
    };
    const pendingParentHeaders = [];
    headersToGroup.forEach((headerToGroup) => {
      const latestPendingParentHeader = [...pendingParentHeaders].reverse()[0];
      const isLeafHeader = headerToGroup.column.depth === headerGroup.depth;
      let column;
      let isPlaceholder = false;
      if (isLeafHeader && headerToGroup.column.parent) {
        column = headerToGroup.column.parent;
      } else {
        column = headerToGroup.column;
        isPlaceholder = true;
      }
      if (latestPendingParentHeader && latestPendingParentHeader?.column === column) {
        latestPendingParentHeader.subHeaders.push(headerToGroup);
      } else {
        const header = createHeader(table, column, {
          id: [headerFamily, depth, column.id, headerToGroup?.id].filter(Boolean).join("_"),
          isPlaceholder,
          placeholderId: isPlaceholder ? `${pendingParentHeaders.filter((d) => d.column === column).length}` : void 0,
          depth,
          index: pendingParentHeaders.length
        });
        header.subHeaders.push(headerToGroup);
        pendingParentHeaders.push(header);
      }
      headerGroup.headers.push(headerToGroup);
      headerToGroup.headerGroup = headerGroup;
    });
    headerGroups.push(headerGroup);
    if (depth > 0) {
      createHeaderGroup(pendingParentHeaders, depth - 1);
    }
  };
  const bottomHeaders = columnsToGroup.map(
    (column, index) => createHeader(table, column, {
      depth: maxDepth,
      index
    })
  );
  createHeaderGroup(bottomHeaders, maxDepth - 1);
  headerGroups.reverse();
  const recurseHeadersForSpans = (headers) => {
    const filteredHeaders = headers.filter(
      (header) => header.column.getIsVisible()
    );
    return filteredHeaders.map((header) => {
      let colSpan = 0;
      let rowSpan = 0;
      let childRowSpans = [0];
      if (header.subHeaders && header.subHeaders.length) {
        childRowSpans = [];
        recurseHeadersForSpans(header.subHeaders).forEach(
          ({ colSpan: childColSpan, rowSpan: childRowSpan }) => {
            colSpan += childColSpan;
            childRowSpans.push(childRowSpan);
          }
        );
      } else {
        colSpan = 1;
      }
      const minChildRowSpan = Math.min(...childRowSpans);
      rowSpan = rowSpan + minChildRowSpan;
      header.colSpan = colSpan;
      header.rowSpan = rowSpan;
      return { colSpan, rowSpan };
    });
  };
  recurseHeadersForSpans(headerGroups[0]?.headers ?? []);
  return headerGroups;
}
export {
  Headers,
  buildHeaderGroups
};
