// @dynatrace/strato-components/esm/tables/DataTable/tanstack-table-core-override/src/core/table.js
import { functionalUpdate, getMemoOptions, memo } from "../utils";
import { createColumn } from "./column";
import { Headers } from "./headers";
import { ColumnFaceting } from "../features/ColumnFaceting";
import { ColumnFiltering } from "../features/ColumnFiltering";
import { ColumnGrouping } from "../features/ColumnGrouping";
import { ColumnOrdering } from "../features/ColumnOrdering";
import { ColumnPinning } from "../features/ColumnPinning";
import { ColumnSizing } from "../features/ColumnSizing";
import { ColumnVisibility } from "../features/ColumnVisibility";
import { GlobalFaceting } from "../features/GlobalFaceting";
import { GlobalFiltering } from "../features/GlobalFiltering";
import { RowExpanding } from "../features/RowExpanding";
import { RowPagination } from "../features/RowPagination";
import { RowPinning } from "../features/RowPinning";
import { RowSelection } from "../features/RowSelection";
import { RowSorting } from "../features/RowSorting";
const builtInFeatures = [
  Headers,
  ColumnVisibility,
  ColumnOrdering,
  ColumnPinning,
  ColumnFaceting,
  ColumnFiltering,
  GlobalFaceting,
  //depends on ColumnFaceting
  GlobalFiltering,
  //depends on ColumnFiltering
  RowSorting,
  ColumnGrouping,
  //depends on RowSorting
  RowExpanding,
  RowPagination,
  RowPinning,
  RowSelection,
  ColumnSizing
];
function createTable(options) {
  if (options.debugAll || options.debugTable) {
    console.info("Creating Table Instance...");
  }
  const _features = [...builtInFeatures, ...options._features ?? []];
  let table = { _features };
  const defaultOptions = table._features.reduce((obj, feature) => {
    return Object.assign(obj, feature.getDefaultOptions?.(table));
  }, {});
  const mergeOptions = (options2) => {
    if (table.options.mergeOptions) {
      return table.options.mergeOptions(defaultOptions, options2);
    }
    return {
      ...defaultOptions,
      ...options2
    };
  };
  const coreInitialState = {};
  let initialState = {
    ...coreInitialState,
    ...options.initialState ?? {}
  };
  table._features.forEach((feature) => {
    initialState = feature.getInitialState?.(initialState) ?? initialState;
  });
  const queued = [];
  let queuedTimeout = false;
  const coreInstance = {
    _features,
    options: {
      ...defaultOptions,
      ...options
    },
    initialState,
    _queue: (cb) => {
      queued.push(cb);
      if (!queuedTimeout) {
        queuedTimeout = true;
        Promise.resolve().then(() => {
          while (queued.length) {
            queued.shift()();
          }
          queuedTimeout = false;
        }).catch(
          (error) => setTimeout(() => {
            throw error;
          })
        );
      }
    },
    reset: () => {
      table.setState(table.initialState);
    },
    setOptions: (updater) => {
      const newOptions = functionalUpdate(updater, table.options);
      table.options = mergeOptions(newOptions);
    },
    getState: () => {
      return table.options.state;
    },
    setState: (updater) => {
      table.options.onStateChange?.(updater);
    },
    _getRowId: (row, index, parent) => table.options.getRowId?.(row, index, parent) ?? `${parent ? [parent.id, index].join(".") : index}`,
    getCoreRowModel: () => {
      if (!table._getCoreRowModel) {
        table._getCoreRowModel = table.options.getCoreRowModel(table);
      }
      return table._getCoreRowModel();
    },
    // The final calls start at the bottom of the model,
    // expanded rows, which then work their way up
    getRowModel: () => {
      return table.getPaginationRowModel();
    },
    //in next version, we should just pass in the row model as the optional 2nd arg
    getRow: (id, searchAll) => {
      let row = (searchAll ? table.getPrePaginationRowModel() : table.getRowModel()).rowsById[id];
      if (!row) {
        row = table.getCoreRowModel().rowsById[id];
        if (!row) {
          if (true) {
            throw new Error(`getRow could not find row with ID: ${id}`);
          }
          throw new Error();
        }
      }
      return row;
    },
    _getDefaultColumnDef: memo(
      () => [table.options.defaultColumn],
      (defaultColumn) => {
        defaultColumn = defaultColumn ?? {};
        return {
          header: (props) => {
            const resolvedColumnDef = props.header.column.columnDef;
            if (resolvedColumnDef.accessorKey) {
              return resolvedColumnDef.accessorKey;
            }
            if (resolvedColumnDef.accessorFn) {
              return resolvedColumnDef.id;
            }
            return null;
          },
          // footer: props => props.header.column.id,
          cell: (props) => props.renderValue()?.toString?.() ?? null,
          ...table._features.reduce((obj, feature) => {
            return Object.assign(obj, feature.getDefaultColumnDef?.());
          }, {}),
          ...defaultColumn
        };
      },
      getMemoOptions(options, "debugColumns", "_getDefaultColumnDef")
    ),
    _getColumnDefs: () => table.options.columns,
    getAllColumns: memo(
      () => [table._getColumnDefs()],
      (columnDefs) => {
        const recurseColumns = (columnDefs2, parent, depth = 0) => {
          return columnDefs2.map((columnDef) => {
            const column = createColumn(table, columnDef, depth, parent);
            const groupingColumnDef = columnDef;
            column.columns = groupingColumnDef.columns ? recurseColumns(groupingColumnDef.columns, column, depth + 1) : [];
            return column;
          });
        };
        return recurseColumns(columnDefs);
      },
      getMemoOptions(options, "debugColumns", "getAllColumns")
    ),
    getAllFlatColumns: memo(
      () => [table.getAllColumns()],
      (allColumns) => {
        return allColumns.flatMap((column) => {
          return column.getFlatColumns();
        });
      },
      getMemoOptions(options, "debugColumns", "getAllFlatColumns")
    ),
    _getAllFlatColumnsById: memo(
      () => [table.getAllFlatColumns()],
      (flatColumns) => {
        return flatColumns.reduce((acc, column) => {
          acc[column.id] = column;
          return acc;
        }, {});
      },
      getMemoOptions(options, "debugColumns", "getAllFlatColumnsById")
    ),
    getAllLeafColumns: memo(
      () => [table.getAllColumns(), table._getOrderColumnsFn()],
      (allColumns, orderColumns) => {
        let leafColumns = allColumns.flatMap(
          (column) => column.getLeafColumns()
        );
        return orderColumns(leafColumns);
      },
      getMemoOptions(options, "debugColumns", "getAllLeafColumns")
    ),
    getColumn: (columnId) => {
      const column = table._getAllFlatColumnsById()[columnId];
      if (!column) {
        console.error(`[Table] Column with id '${columnId}' does not exist.`);
      }
      return column;
    }
  };
  Object.assign(table, coreInstance);
  for (let index = 0; index < table._features.length; index++) {
    const feature = table._features[index];
    feature?.createTable?.(table);
  }
  return table;
}
export {
  createTable
};
