// @dynatrace/strato-components/esm/charts/xy-chart/components/toolbar/DownloadDataButton.js
import { jsx } from "react/jsx-runtime";
import { useRef } from "react";
import { useIntl } from "react-intl";
import { DownloadIcon } from "@dynatrace/strato-icons";
import { ChartToolbarRenderer } from "../../../core/components/toolbar/ChartToolbarRenderer.js";
import { useA11yToolbarTabIndex } from "../../../core/hooks/useA11yToolbarTabIndex.js";
import { useDownload } from "../../hooks/useDownload.js";
import { useXYChartData } from "../../hooks/useXYChartData.js";
import { XYCHART_MESSAGES } from "../../messages/messages.js";
import { buildAndDownloadCsv } from "../../utils/build-and-download-csv.js";
const DownloadDataButton = () => {
  const downloadDataButtonRef = useRef(null);
  const tabIndex = useA11yToolbarTabIndex("download");
  const intl = useIntl();
  const { downloadDataCallback } = useDownload();
  const { seriesWithDatapoints, datapointMetadataMap, seriesToMetadata } = useXYChartData();
  const triggerDownload = () => {
    buildAndDownloadCsv(
      seriesWithDatapoints,
      seriesToMetadata,
      datapointMetadataMap,
      downloadDataCallback
    );
  };
  const handleDownloadDataKeyboard = (event) => {
    if (event.key === "Enter") {
      triggerDownload();
    }
  };
  const handleDownloadDataToolbar = () => {
    triggerDownload();
  };
  return /* @__PURE__ */ jsx(
    ChartToolbarRenderer.Control,
    {
      ref: downloadDataButtonRef,
      prefixIcon: /* @__PURE__ */ jsx(DownloadIcon, {}),
      tooltipText: intl.formatMessage(XYCHART_MESSAGES.downloadAsCsv),
      onKeyDown: handleDownloadDataKeyboard,
      onClick: handleDownloadDataToolbar,
      tabIndex,
      divider: "left",
      "data-dt-name": "ToolbarDownloadData"
    }
  );
};
DownloadDataButton["displayName"] = "DownloadDataButton";
export {
  DownloadDataButton
};
