// @dynatrace/strato-components/esm/charts/xy-chart/components/toolbar/PanButton.js
import { jsx } from "react/jsx-runtime";
import { useRef } from "react";
import { useIntl } from "react-intl";
import { PanHorizontalIcon } from "@dynatrace/strato-icons";
import { allowWhenKeyIsEnter } from "./utils.js";
import { KeyboardControls } from "../../../core/components/interactions-overlay/constants.js";
import { INTERACTION_MODE } from "../../../core/components/interactions-overlay/types.js";
import { useSynchronizedSelection } from "../../../core/components/shared-chart-interactions/hooks/useSynchronizedSelection.js";
import { ChartToolbarRenderer } from "../../../core/components/toolbar/ChartToolbarRenderer.js";
import { TOOLBAR_MESSAGES } from "../../../core/components/toolbar/messages.js";
import { useA11yToolbarTabIndex } from "../../../core/hooks/useA11yToolbarTabIndex.js";
import { getKeyShortcutByAgent } from "../../../core/utils/get-key-shortcut-by-agent.js";
import { XYCHART_MESSAGES } from "../../messages/messages.js";
import { reduceInteractionMode } from "../../store/reduce-interaction-mode.js";
import { reduceExploreCancelAll } from "../../store/reducer-explorer.js";
import {
  useInteractionModeState,
  useSetState,
  useZoomState
} from "../../store/selectors.js";
const PanButton = () => {
  const setState = useSetState();
  const panButtonRef = useRef(null);
  const intl = useIntl();
  const tabIndex = useA11yToolbarTabIndex(INTERACTION_MODE.Pan);
  const { zoomOutEnabled } = useZoomState();
  const interactionMode = useInteractionModeState();
  const { resetSelectionState } = useSynchronizedSelection();
  const isSelected = interactionMode === INTERACTION_MODE.Pan;
  const action = () => {
    resetSelectionState?.();
    setState(reduceExploreCancelAll);
    setState(reduceInteractionMode(INTERACTION_MODE.Pan));
  };
  const tooltipKeyShortcut = `${getKeyShortcutByAgent(
    "ALT"
  )} ${intl.formatMessage(TOOLBAR_MESSAGES.or)} ${KeyboardControls.KEYBOARD_PAN_MODE_SHORTCUT_ALTERNATIVE}`;
  return /* @__PURE__ */ jsx(
    ChartToolbarRenderer.Control,
    {
      ref: panButtonRef,
      prefixIcon: /* @__PURE__ */ jsx(PanHorizontalIcon, {}),
      tooltipText: intl.formatMessage(XYCHART_MESSAGES.zoomModePan),
      tooltipKeyShortcut,
      selected: isSelected,
      onKeyDown: allowWhenKeyIsEnter(action),
      onClick: action,
      tabIndex,
      disabled: !zoomOutEnabled,
      placement: "menu",
      "data-dt-name": "ToolbarPanMode"
    }
  );
};
PanButton["displayName"] = "PanButton";
export {
  PanButton
};
