// @dynatrace/strato-components/esm/charts/xy-chart/components/toolbar/XYChartToolbar.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { useIntl } from "react-intl";
import { DownloadDataButton } from "./DownloadDataButton.js";
import { ExploreButton } from "./ExploreButton.js";
import { PanButton } from "./PanButton.js";
import { ResetButton } from "./ResetButton.js";
import { ZoomInOutButtons } from "./ZoomInOutButtons.js";
import { ZoomModeButton } from "./ZoomModeButton.js";
import { useBehavioralTrackingProps as _useBehavioralTrackingProps } from "../../../../core/hooks/useBehavioralTrackingProps.js";
import { ChartToolbarRenderer } from "../../../core/components/toolbar/ChartToolbarRenderer.js";
import { getMenuTriggerDisplayMode } from "../../../core/components/toolbar/utils/get-menu-trigger-display-mode.js";
import { getMenuTriggerIcon } from "../../../core/components/toolbar/utils/get-menu-trigger-icon.js";
import { ChartOverlaysElevation } from "../../../core/constants.js";
import { getA11yToolbarConfig } from "../../../core/utils/zoom.js";
import { defaultToolbar } from "../../constants/defaults.js";
import { useDownload } from "../../hooks/useDownload.js";
import { useToolbar } from "../../hooks/useToolbar.js";
import { useXYChartData } from "../../hooks/useXYChartData.js";
import { useXYChartSelectionAreaRef } from "../../hooks/useXYChartSelectionAreaRef.js";
import { XYCHART_MESSAGES } from "../../messages/messages.js";
import { useInteractionModeState } from "../../store/selectors.js";
import { buildAndDownloadCsv } from "../../utils/build-and-download-csv.js";
const XYChartToolbar = ({
  onMouseEnter,
  onMouseLeave,
  handlerRef
}) => {
  const toolbarContext = useToolbar();
  const { collapsed, placement, hidden, intents } = toolbarContext;
  const { downloadDataCallback } = useDownload();
  const interactionMode = useInteractionModeState();
  const { axesMetadata } = useXYChartData();
  const intl = useIntl();
  const selectionAreaRef = useXYChartSelectionAreaRef();
  const isZoomEnabled = axesMetadata.isZoomEnabled;
  const isPanEnabled = axesMetadata.isPanEnabled;
  const exploreBtnPlace = isZoomEnabled || isPanEnabled ? "menu" : "auto";
  const isDownloadEnabled = true;
  const initialA11yState = getA11yToolbarConfig({
    zoom: { enabled: isZoomEnabled },
    pan: { enabled: isPanEnabled },
    download: { enabled: isDownloadEnabled },
    intent: !!intents
  });
  const menuIcon = getMenuTriggerIcon(interactionMode, isZoomEnabled);
  const menuLabel = !isPanEnabled && !isZoomEnabled ? intl.formatMessage(XYCHART_MESSAGES.zoomModeExplore) : void 0;
  const menuDisplayMode = getMenuTriggerDisplayMode({
    isPanEnabled,
    isZoomEnabled,
    isDownloadEnabled
  });
  const { seriesWithDatapoints, datapointMetadataMap, seriesToMetadata } = useXYChartData();
  const handleDownloadDataToolbar = () => {
    buildAndDownloadCsv(
      seriesWithDatapoints,
      seriesToMetadata,
      datapointMetadataMap,
      downloadDataCallback
    );
  };
  const showDefaultToolbarDownloadButton = isDownloadEnabled && !intents;
  const showMenuToolbarButton = isDownloadEnabled && !!intents?.length;
  const onClose = () => selectionAreaRef ? selectionAreaRef.current?.focus() : void 0;
  const behaviorTrackingProps = _useBehavioralTrackingProps(toolbarContext)[0];
  return !hidden ? /* @__PURE__ */ jsxs(
    ChartToolbarRenderer,
    {
      visible: false,
      visibleWhenMouseInBounds: true,
      handlerRef,
      placement: placement ?? defaultToolbar.placement,
      draggable: true,
      collapsed,
      initialA11yState,
      zIndex: ChartOverlaysElevation.toolbar,
      onMouseEnter,
      onMouseLeave,
      menuIcon,
      menuTriggerLabel: menuLabel,
      menuDisplayMode,
      isDownloadEnabled: showMenuToolbarButton,
      intents,
      handleDownloadDataToolbar,
      onClose,
      ...behaviorTrackingProps,
      children: [
        /* @__PURE__ */ jsxs(Fragment, { children: [
          /* @__PURE__ */ jsx(ExploreButton, { placement: exploreBtnPlace }),
          isZoomEnabled && /* @__PURE__ */ jsx(ZoomModeButton, {}),
          isPanEnabled && /* @__PURE__ */ jsx(PanButton, {})
        ] }),
        isZoomEnabled && /* @__PURE__ */ jsx(ZoomInOutButtons, {}),
        isZoomEnabled && /* @__PURE__ */ jsx(ResetButton, {}),
        showDefaultToolbarDownloadButton && /* @__PURE__ */ jsx(DownloadDataButton, {})
      ]
    }
  ) : null;
};
XYChartToolbar["displayName"] = "XYChartToolbar";
export {
  XYChartToolbar
};
