// @dynatrace/strato-components/esm/forms/spin-buttons/spin-button-group/SpinButtonGroup.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import { useNavigateSpinButton } from "./hooks/useNavigateSpinButton.js";
import { useValueOverflow } from "./hooks/useValueOverflow.js";
import { SpinButtonGroupSeparator } from "./SpinButtonGroupSeparator.js";
import { useBehavioralTrackingProps } from "../../../core/hooks/useBehavioralTrackingProps.js";
import { NumberSpinButton } from "../number-spin-button/NumberSpinButton.js";
import { spinButtonGroupCSS } from "../SpinButton.sty.js";
import { TextSpinButton } from "../text-spin-button/TextSpinButton.js";
const InternalSpinButtonGroup = ({
  children,
  valueOverflow = false,
  onLeadingEdge,
  onTrailingEdge,
  onTrailingOverflow,
  className: consumerClassName,
  style: consumerStyle,
  "data-testid": dataTestId,
  "data-dtrum-mask": dataDtrumMask,
  "data-dtrum-allow": dataDtrumAllow,
  ...remainingProps
}) => {
  const [behavioralTrackingProps] = useBehavioralTrackingProps(remainingProps);
  const { clonedChildren, spinButtonRefs, triggerRefs } = useValueOverflow(
    children,
    valueOverflow,
    onTrailingOverflow,
    onTrailingEdge
  );
  const { onKeyDownHandler } = useNavigateSpinButton(
    spinButtonRefs,
    triggerRefs,
    onLeadingEdge,
    onTrailingEdge
  );
  return /* @__PURE__ */ jsx(
    "span",
    {
      className: clsx(spinButtonGroupCSS, consumerClassName),
      "data-valuetype": "iso8601",
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...behavioralTrackingProps,
      onKeyDown: onKeyDownHandler,
      children: clonedChildren
    }
  );
};
const SpinButtonGroup = Object.assign(InternalSpinButtonGroup, {
  Number: NumberSpinButton,
  Text: TextSpinButton,
  Separator: SpinButtonGroupSeparator
});
SpinButtonGroup.displayName = "_SpinButtonGroup";
export {
  SpinButtonGroup
};
