// @dynatrace/strato-components/esm/charts/timeseries/hooks/useBuildFilteredMetadata.js
import { isNil, uniq } from "lodash-es";
import { useMinMaxXAxis } from "./useMinMaxXAxis.js";
import { useXAxisConfig } from "./useXAxisConfig.js";
import { useYAxisConfig } from "./useYAxisConfig.js";
import { getThresholdValue } from "../../core/utils/apply-gap-policy.js";
import { getDatapointIsGap } from "../../core/utils/get-data-point-is-gap.js";
import { safeGet } from "../../core/utils/safe-get.js";
import {
  getCumulativeValue
} from "../components/displays/shared/utils/datapoint-stacking.js";
import { getDatapointValue } from "../utils/get-datapoint-value.js";
import { getRelativeConf } from "../utils/get-relative-conf.js";
import { getSanitizedValueRepresentation } from "../utils/get-sanitized-value-representation.js";
import { isStacked } from "../utils/is-stacked.js";
import { getDatapointTime } from "../utils/timeseries.js";
function useBuildFilteredMetadata(data, valueRepPerVariant, seriesMap, units) {
  const xAxisConfig = useXAxisConfig();
  const yAxisConfig = useYAxisConfig();
  const { min, max } = useMinMaxXAxis();
  const datapointsMap = /* @__PURE__ */ new WeakMap();
  const sanitizedValueRepPerVariant = getSanitizedValueRepresentation(
    data,
    valueRepPerVariant
  );
  const filteredTimeseries = { ...data };
  const filteredUnits = [];
  const originalFilteredUnits = [];
  let relativeGroupedValues;
  for (const variant of data.shapePriority) {
    const cumulativeStackedPoints = /* @__PURE__ */ new Map();
    const lastSignStacked = /* @__PURE__ */ new Map();
    const isBand = variant === "band";
    const variantSeries = data[variant];
    relativeGroupedValues = getRelativeConf(
      variantSeries,
      variant,
      sanitizedValueRepPerVariant,
      !isStacked(variant) ? relativeGroupedValues : void 0
    );
    const newSeries = [];
    for (let i = 0; i < variantSeries.length; i++) {
      const series = variantSeries[i];
      const seriesMetadata = safeGet(seriesMap, series);
      const { id, color, gapPolicy, pointsDisplay } = seriesMetadata;
      const unit = relativeGroupedValues ? "%" : series.unit;
      const originalUnit = series.unit;
      const metadata = {
        id,
        color,
        unit,
        originalUnit,
        gapPolicy,
        pointsDisplay
      };
      if (units.length === 1 && yAxisConfig.right && !yAxisConfig.left) {
        filteredUnits.push(void 0);
      }
      filteredUnits.push(unit);
      originalFilteredUnits.push(originalUnit);
      seriesMap.set(series, metadata);
      const thresholdGap = getThresholdValue(gapPolicy);
      const newDatapoints = [];
      for (let j = 0; j < series.datapoints.length; j++) {
        const datapoint = series.datapoints[j];
        if (xAxisConfig && (xAxisConfig.min || xAxisConfig.max)) {
          const dataPointTimestamp = getDatapointTime(
            datapoint.start,
            datapoint.end,
            datapoint.center
          );
          const dataPointOutOfBounds = dataPointTimestamp < +min || dataPointTimestamp > +max;
          if (dataPointOutOfBounds) {
            continue;
          }
        }
        const nextDatapoint = series.datapoints[j + 1];
        const { absolute, value, y0, absoluteY0, y1, absoluteY1 } = getDatapointValue(
          datapoint,
          variant,
          originalUnit,
          relativeGroupedValues
        );
        const isBandPointUndefined = isBand && (isNil(y0) || isNil(y1));
        const isPointUndefined = !isBand && isNil(value);
        if (isBandPointUndefined || isPointUndefined) {
          continue;
        }
        const isGap = gapPolicy !== "connect" && getDatapointIsGap(datapoint, nextDatapoint, thresholdGap);
        let stackedSign = value ?? 0;
        const key = datapoint.start.getTime().toString() + datapoint.end?.getTime();
        if (value === 0) {
          const prevSign = lastSignStacked.get(key) ?? 0;
          stackedSign = prevSign;
        }
        const cumulativeStack = isStacked(variant) ? getCumulativeValue(
          cumulativeStackedPoints,
          datapoint,
          relativeGroupedValues ? originalUnit : unit,
          id,
          value,
          stackedSign
        ) : void 0;
        const datapointMetadata = {
          start: datapoint.start,
          end: datapoint.end,
          center: datapoint.center,
          seriesId: id,
          color,
          unit,
          isGap: false,
          absoluteValue: absolute,
          value,
          variant,
          y0,
          y1,
          absoluteY0,
          absoluteY1,
          cumulativeValue: cumulativeStack?.cumulativeValue ?? 0,
          seriesName: series.name,
          originalUnit
        };
        if (isStacked(variant) && cumulativeStack) {
          cumulativeStackedPoints.set(cumulativeStack.key, {
            seriesId: id,
            datapointsCumulative: cumulativeStack.datapointsCumulative,
            baseCumulative: cumulativeStack.baseCumulative
          });
          lastSignStacked.set(key, stackedSign);
        }
        datapointsMap.set(datapoint, datapointMetadata);
        newDatapoints.push(datapoint);
        if (isGap) {
          const metadataGappedDatapoint = {
            ...datapointMetadata,
            isGap: true,
            value: Infinity
          };
          const gappedDatapoint = {
            start: datapoint.end ?? datapoint.start,
            end: nextDatapoint.start,
            isGap: true,
            value: Infinity
          };
          datapointsMap.set(gappedDatapoint, metadataGappedDatapoint);
          newDatapoints.push(gappedDatapoint);
        }
      }
      const { datapoints, ...spreadSeries } = series;
      const newSingleSeries = { ...spreadSeries, datapoints: newDatapoints };
      newSeries.push(newSingleSeries);
      seriesMap.set(newSingleSeries, metadata);
    }
    filteredTimeseries[variant] = newSeries;
  }
  return {
    seriesMap,
    datapointsMap,
    filteredUnits: uniq(filteredUnits),
    filteredTimeseries,
    originalFilteredUnits: uniq(originalFilteredUnits)
  };
}
export {
  useBuildFilteredMetadata
};
