// @dynatrace/strato-components/forms/calendar/hooks/useMonth.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useMonth_exports = {};
__export(useMonth_exports, {
  useMonth: () => useMonth
});
module.exports = __toCommonJS(useMonth_exports);
var import_tz = require("@date-fns/tz");
var import_date_fns = require("date-fns");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_units = require("@dynatrace-sdk/units");
var import_user_preferences = require("@dynatrace-sdk/user-preferences");
var import_useCalendarContext = require("./useCalendarContext.js");
var import_useFocus = require("./useFocus.js");
var import_helpers = require("../../../core/time/helpers.js");
var import_useTimeframeStore = require("../../../core/time/timeframeStore/useTimeframeStore.js");
var import_validation = require("../../../core/time/validation.js");
function useMonth({ date, itemRef }) {
  const { state, dispatch } = (0, import_useCalendarContext.useCalendarContext)();
  const configuredTimezone = (0, import_user_preferences.getTimezone)();
  const value = (0, import_useTimeframeStore.useCalendarStore)((store) => store.calendarValue);
  const min = (0, import_useTimeframeStore.useCalendarStore)((store) => store.min);
  const max = (0, import_useTimeframeStore.useCalendarStore)((store) => store.max);
  const intl = (0, import_react_intl.useIntl)();
  const ariaLabel = (0, import_react.useMemo)(() => {
    const formatMessage = intl.formatMessage({
      defaultMessage: "Select month",
      id: "aRc3KoQgfpti2E9M",
      description: "Aria label to add to each month button"
    });
    return `${formatMessage} ${(0, import_units.formatDate)(date, {
      month: "long"
    })}`;
  }, [intl, date]);
  const isCurrent = (0, import_react.useMemo)(
    () => (0, import_date_fns.isSameMonth)(date, state.today),
    [date, state.today]
  );
  const isFocused = (0, import_react.useMemo)(
    () => (0, import_date_fns.isSameMonth)(date, state.activeDate),
    [date, state.activeDate]
  );
  const isSelected = (0, import_react.useMemo)(() => {
    if ((0, import_helpers.isTimeframe)(value, state.type) && value) {
      const { from, to } = value;
      const fromDate = from ? new import_tz.TZDateMini(from.absoluteDate, configuredTimezone) : null;
      const toDate = to ? new import_tz.TZDateMini(to.absoluteDate, configuredTimezone) : null;
      return !!(fromDate && (0, import_date_fns.isSameMonth)(fromDate, date) || toDate && (0, import_date_fns.isSameMonth)(toDate, date));
    }
    if ((0, import_helpers.isTimeValue)(value)) {
      const selectedDate = value ? (0, import_date_fns.startOfMonth)(new import_tz.TZDateMini(value.absoluteDate, configuredTimezone)) : null;
      if (!selectedDate) {
        return false;
      }
      return (0, import_date_fns.isSameMonth)(selectedDate, date);
    }
  }, [configuredTimezone, value, state.type, date]);
  const { tabIndex } = (0, import_useFocus.useFocus)({ isFocused, itemRef });
  const isDisabled = (0, import_react.useMemo)(() => {
    const startDay = (0, import_date_fns.startOfMonth)(date);
    const endDay = (0, import_date_fns.endOfMonth)(date);
    return !!((0, import_validation.isOutOfBounds)(startDay, void 0, max) || (0, import_validation.isOutOfBounds)(endDay, min));
  }, [min, max, date]);
  function onClick() {
    dispatch({
      type: "activeDateChange",
      payload: { activeDate: date, changeToView: "day" }
    });
    dispatch({ type: "shouldUpdateFocusFromKeydown", payload: false });
  }
  return {
    ariaLabel,
    tabIndex,
    isCurrent,
    isSelected,
    isDisabled,
    onClick
  };
}
