// @dynatrace/strato-components/forms/calendar/hooks/useYear.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useYear_exports = {};
__export(useYear_exports, {
  useYear: () => useYear
});
module.exports = __toCommonJS(useYear_exports);
var import_tz = require("@date-fns/tz");
var import_date_fns = require("date-fns");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_user_preferences = require("@dynatrace-sdk/user-preferences");
var import_useCalendarContext = require("./useCalendarContext.js");
var import_useFocus = require("./useFocus.js");
var import_helpers = require("../../../core/time/helpers.js");
var import_useTimeframeStore = require("../../../core/time/timeframeStore/useTimeframeStore.js");
var import_validation = require("../../../core/time/validation.js");
function useYear({ date, itemRef }) {
  const { state, dispatch } = (0, import_useCalendarContext.useCalendarContext)();
  const configuredTimezone = (0, import_user_preferences.getTimezone)();
  const value = (0, import_useTimeframeStore.useCalendarStore)((store) => store.calendarValue);
  const min = (0, import_useTimeframeStore.useCalendarStore)((store) => store.min);
  const max = (0, import_useTimeframeStore.useCalendarStore)((store) => store.max);
  const intl = (0, import_react_intl.useIntl)();
  const ariaLabel = (0, import_react.useMemo)(() => {
    const formatMessage = intl.formatMessage({
      defaultMessage: "Select year",
      id: "pRBtzsvqzp1/QqAV",
      description: "Aria label to add to each year button"
    });
    return `${formatMessage} ${date.getFullYear().toString()}`;
  }, [date, intl]);
  const isCurrent = (0, import_react.useMemo)(
    () => (0, import_date_fns.isSameYear)(date, state.today),
    [date, state.today]
  );
  const isFocused = (0, import_react.useMemo)(
    () => (0, import_date_fns.isSameYear)(date, state.activeDate),
    [state.activeDate, date]
  );
  const isSelected = (0, import_react.useMemo)(() => {
    const compareYear = date.getFullYear();
    if (!value) {
      return false;
    }
    if ((0, import_helpers.isTimeValue)(value)) {
      const selectedYear = value ? new import_tz.TZDateMini(value.absoluteDate, configuredTimezone).getFullYear() : null;
      return compareYear === selectedYear;
    }
    const { from, to } = value;
    const fromYear = from ? new import_tz.TZDateMini(from.absoluteDate, configuredTimezone).getFullYear() : null;
    const toYear = to ? new import_tz.TZDateMini(to.absoluteDate, configuredTimezone).getFullYear() : null;
    return fromYear === compareYear || toYear === compareYear;
  }, [configuredTimezone, value, date]);
  const isDisabled = (0, import_react.useMemo)(() => {
    const startDay = (0, import_date_fns.startOfYear)(date);
    const endDay = (0, import_date_fns.endOfYear)(date);
    return !!((0, import_validation.isOutOfBounds)(startDay, void 0, max) || (0, import_validation.isOutOfBounds)(endDay, min));
  }, [min, max, date]);
  const { tabIndex } = (0, import_useFocus.useFocus)({ isFocused, itemRef });
  function onClick() {
    dispatch({
      type: "activeDateChange",
      payload: { activeDate: date, changeToView: "month" }
    });
    dispatch({ type: "shouldUpdateFocusFromKeydown", payload: false });
  }
  return {
    ariaLabel,
    tabIndex,
    isCurrent,
    isSelected,
    isDisabled,
    onClick
  };
}
