// @dynatrace/strato-components/esm/forms/number-input/NumberInputBase.js
import { jsx, jsxs } from "react/jsx-runtime";
import { useNumberField } from "@react-aria/numberfield";
import { useNumberFieldState } from "@react-stately/numberfield";
import clsx from "clsx";
import {
  forwardRef,
  useImperativeHandle,
  useMemo,
  useRef
} from "react";
import { useIntl } from "react-intl";
import { useAriaLabelingProps } from "../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useControllableState } from "../../core/hooks/useControllableState.js";
import { useFormControlState } from "../../core/hooks/useFormControlState.js";
import { useFocusRing } from "../../core/styles/useFocusRing.js";
import { mergeProps as _mergeProps } from "../../core/utils/merge-props.js";
import { Flex } from "../../layouts/flex/Flex.js";
import { inputGroupClassName } from "../../layouts/input-group/InputGroup.sty.js";
import * as styles from "../_base-input/BaseInput.sty.js";
import { InternalBaseItem as BaseItem } from "../forms-core/components/form-field-messages/FormFieldMessageBaseItem.js";
import { useSyncedFormControlProps } from "../forms-core/hooks/useSyncedFormControlProps.js";
import { getSafeRegionalFormat } from "../forms-core/time/helpers.js";
import { autofillCSS } from "../hooks/useAutofillEvent.sty.js";
import { useAutofillEvent } from "../hooks/useAutofillEvent.js";
import { useInputHint } from "../hooks/useInputHint.js";
const NumberInputBase = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const intl = useIntl();
  const defaultPlaceholder = intl.formatMessage({
    defaultMessage: "Add number",
    id: "rrkUQoVNb46Jbd/8",
    description: "Default placeholder for the number input."
  });
  const {
    id: propId,
    defaultValue,
    value,
    name,
    onChange,
    disabled: propDisabled,
    readOnly: propReadOnly,
    required: propRequired,
    min: minValueNumberLike,
    max: maxValueNumberLike,
    step,
    placeholder = defaultPlaceholder,
    controlState,
    onKeyUp,
    onKeyDown,
    onBlur,
    onFocus,
    onFocusChange,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "aria-disabled": ariaDisabledProp,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
  const componentName = restProps["data-dt-component"];
  const { "aria-labelledby": ariaLabelledBy, ...remainingAriaProps } = useAriaLabelingProps(restProps);
  const locale = useMemo(() => getSafeRegionalFormat(), []);
  const minValue = !isNaN(Number(minValueNumberLike)) ? Number(minValueNumberLike) : void 0;
  const maxValue = !isNaN(Number(maxValueNumberLike)) ? Number(maxValueNumberLike) : void 0;
  const numberInputValue = value === null || value === void 0 ? NaN : value;
  const {
    disabled,
    required,
    id: contextId,
    ariaDisabled,
    readOnly: isReadOnly
  } = useSyncedFormControlProps({
    propDisabled,
    propRequired,
    propId,
    propAriaDisabled: ariaDisabledProp,
    propReadOnly
  });
  const disabledOrAriaDisabled = disabled || ariaDisabled || false;
  const state = useNumberFieldState({
    value: numberInputValue,
    onChange,
    placeholder,
    minValue,
    maxValue,
    step,
    isDisabled: disabled,
    isReadOnly,
    isRequired: required,
    locale
  });
  const inputRef = useRef(null);
  const wrapperRef = useRef(null);
  useImperativeHandle(
    forwardedRef,
    () => ({
      element: wrapperRef.current,
      inputRef: inputRef.current,
      focus() {
        inputRef.current?.focus();
      }
    }),
    []
  );
  const { inputProps } = useNumberField(
    {
      value: numberInputValue,
      onChange,
      id: contextId,
      placeholder,
      minValue,
      maxValue,
      step,
      isDisabled: disabled,
      isReadOnly,
      isRequired: required,
      onKeyUp,
      onKeyDown,
      onBlur,
      onFocus,
      onFocusChange,
      // we pass the aria-labelledBy only if there's no aria-label, because the useNumberField would add aria-labelledby
      // with 2 ids and append the input text to its aria-label otherwise, breaking the getByLabelText selectors in tests otherwise
      "aria-labelledby": remainingAriaProps["aria-label"] ? void 0 : ariaLabelledBy,
      ...remainingAriaProps
      // passing the aria props here prevents unnecessary react-aria warnings when aria-labelling props are already set
    },
    state,
    // eslint-disable-next-line @typescript-eslint/prefer-ts-expect-error, @typescript-eslint/ban-ts-comment
    // @ts-ignore React 19 - proper overlap for the ref is no longer there. It should be fine, but the Numberfield will be reworked anyways
    inputRef
  );
  let inputDefaultValue;
  if (defaultValue !== null && defaultValue !== void 0) {
    inputDefaultValue = defaultValue.toString();
  } else {
    inputDefaultValue = value === void 0 && !onChange ? "" : void 0;
  }
  const {
    value: inputPropsValue,
    type,
    id,
    readOnly = false,
    role,
    onKeyDown: onKeyDownReactAria,
    ...remainingInputProps
  } = inputProps;
  const [inputValue, setInputValue] = useControllableState({
    defaultValue: inputDefaultValue,
    value: inputPropsValue?.toString()
  });
  const [hasError, hint] = useFormControlState(controlState);
  const { inputProps: inputHintAriaProps, hintProps } = useInputHint({
    hint,
    hasError
  });
  const { isAutoFilled, onAnimationStart } = useAutofillEvent();
  const handleInputChange = (event) => {
    remainingInputProps?.onChange?.(event);
    setInputValue(event.target.value);
  };
  const { focusProps, focusClassName } = useFocusRing({
    color: hasError && !(disabled || ariaDisabled) ? "critical" : "neutral",
    ignoreModality: true,
    disabled
  });
  return /* @__PURE__ */ jsxs(
    Flex,
    {
      gap: 4,
      flexDirection: "column",
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ref: wrapperRef,
      className: consumerClassName,
      style: consumerStyle,
      ...behavioralTrackingProps,
      "data-dt-component": componentName,
      children: [
        /* @__PURE__ */ jsx(
          Flex,
          {
            gap: 2,
            alignItems: "center",
            className: clsx(styles.wrapper, styles.variant["default"], {
              [styles.hasError]: hasError,
              [styles.disabled]: disabledOrAriaDisabled,
              [styles.readOnly]: readOnly
            }),
            children: /* @__PURE__ */ jsx(
              "div",
              {
                className: clsx(
                  styles.container,
                  inputGroupClassName,
                  focusClassName,
                  {
                    [styles.isAutoFilled]: isAutoFilled
                  }
                ),
                children: /* @__PURE__ */ jsx(
                  "input",
                  {
                    id,
                    ref: inputRef,
                    type: "text",
                    role,
                    onChange: handleInputChange,
                    onAnimationStart,
                    disabled,
                    className: clsx(styles.input, autofillCSS),
                    readOnly,
                    ..._mergeProps(
                      remainingInputProps,
                      restProps,
                      focusProps,
                      inputHintAriaProps
                    ),
                    name,
                    required,
                    value: inputValue,
                    "aria-disabled": ariaDisabled,
                    onKeyDown: (e) => {
                      if (e.key === "Home" && minValue === void 0 || e.key === "End" && maxValue === void 0) {
                        onKeyDown?.(e);
                        return;
                      }
                      onKeyDownReactAria?.(e);
                    }
                  }
                )
              }
            )
          }
        ),
        hint && /* @__PURE__ */ jsx(
          BaseItem,
          {
            variant: hasError ? "error" : "hint",
            disabled: disabled || ariaDisabled,
            ...hintProps,
            children: hint
          }
        )
      ]
    }
  );
});
NumberInputBase.displayName = "NumberInputBase";
export {
  NumberInputBase
};
