// @dynatrace/strato-components/filters/filter-field/contexts/FilterFieldContext.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var FilterFieldContext_exports = {};
__export(FilterFieldContext_exports, {
  FilterFieldContext: () => FilterFieldContext
});
module.exports = __toCommonJS(FilterFieldContext_exports);
var import_react = require("react");
var import_suggestions_config = require("../suggestions/suggestions-config.js");
const FilterFieldContext = (0, import_react.createContext)({
  applySuggestion: () => null,
  dispatch: () => null,
  autoSuggestions: {},
  defaultSuggestionsCount: import_suggestions_config.DEFAULT_SUGGESTIONS_COUNT,
  recentSuggestionValues: [],
  pinnedSuggestionValues: [],
  logicalOperatorSuggestionValues: [],
  pastedContentSuggestionValues: [],
  searchSuggestionValues: [],
  fallbackKeySuggestionValues: [],
  statementSuggestionValues: [],
  suggestionsOverlayRef: null,
  triggerRef: null,
  filterNamespace: null,
  editorView: null,
  handleNextSuggestion: () => null
});
