// @dynatrace/strato-components/filters/filter-field/contexts/SavedAndRecentQueryClientContext.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SavedAndRecentQueryClientContext_exports = {};
__export(SavedAndRecentQueryClientContext_exports, {
  RecentAndSavedFiltersQueryClientContext: () => RecentAndSavedFiltersQueryClientContext
});
module.exports = __toCommonJS(SavedAndRecentQueryClientContext_exports);
var import_react_query = require("@tanstack/react-query");
var import_react = require("react");
const queryClient = new import_react_query.QueryClient({
  defaultOptions: {
    queries: {
      retry: false,
      // Data only changes via explicit mutations — we can set the staleTime to infinity to prevent unnecessary refetches.
      staleTime: Infinity
    },
    mutations: {
      retry: false
    }
  }
});
const RecentAndSavedFiltersQueryClientContext = (0, import_react.createContext)(queryClient);
