// @dynatrace/strato-components/esm/charts/timeseries/components/displays/band/accessors.js
import { safeGet } from "../../../../core/utils/safe-get.js";
import { getDatapointTime } from "../../../utils/timeseries.js";
const definedAccessor = (dataPoint, datapointsMap) => !safeGet(datapointsMap, dataPoint).isGap;
const xAxisAccessor = (xScale, datapointsMap) => (dataPoint) => {
  const { start, end, center } = safeGet(datapointsMap, dataPoint);
  return xScale(getDatapointTime(start, end, center)) ?? 0;
};
const y0AxisAccessor = (scale, dataPointsMap) => (dataPoint) => {
  const dataPointMetadata = safeGet(dataPointsMap, dataPoint);
  const value = dataPointMetadata.y0 ?? 0;
  return scale(value);
};
const y1AxisAccessor = (scale, dataPointsMap) => (dataPoint) => {
  const dataPointMetadata = safeGet(dataPointsMap, dataPoint);
  const value = dataPointMetadata.y1 ?? 0;
  return scale(value);
};
function y1(yScaleByUnit, seriesMetadata, datapointsMap) {
  return (dataPoint) => y1AxisAccessor(
    yScaleByUnit(seriesMetadata.unit),
    datapointsMap
  )(dataPoint) ?? 0;
}
function y0(yScaleByUnit, seriesMetadata, datapointsMap) {
  return (dataPoint) => y0AxisAccessor(
    yScaleByUnit(seriesMetadata.unit),
    datapointsMap
  )(dataPoint) ?? 0;
}
export {
  definedAccessor,
  xAxisAccessor,
  y0,
  y0AxisAccessor,
  y1,
  y1AxisAccessor
};
