// @dynatrace/strato-components/esm/charts/timeseries/components/displays/band/BandChartDisplay.js
import { Fragment, jsx } from "react/jsx-runtime";
import { useMemo } from "react";
import { definedAccessor, xAxisAccessor as x, y0, y1 } from "./accessors.js";
import { getBandStyle } from "./band-style.js";
import { AreaCanvas } from "../../../../core/components/canvas/components/AreaCanvas.js";
import { useGridPlotSize } from "../../../../core/components/chart-grid/hooks/useGridPlotSize.js";
import { DEFAULT_STROKE_WIDTH_BAND_CHART } from "../../../../core/constants.js";
import { safeGet } from "../../../../core/utils/safe-get.js";
import { useSeriesCurve } from "../../../hooks/useSeriesCurve.js";
import { useTimeseriesFilteredData } from "../../../hooks/useTimeseriesFilteredData.js";
import { useYScales } from "../../../hooks/useYScales.js";
import { getYScaleByUnit } from "../../../utils/yAxis.js";
const BandChartDisplay = ({
  series: timeseries,
  xScale
}) => {
  const yScale = useYScales();
  const { width, height } = useGridPlotSize();
  const { seriesMap, datapointsMap } = useTimeseriesFilteredData();
  const curve = useSeriesCurve();
  const bandSeries = useMemo(() => [...timeseries].reverse(), [timeseries]);
  const yScaleByUnit = (unit) => getYScaleByUnit(unit, yScale);
  const displayScatterPlot = (_datapoints, _index) => false;
  return /* @__PURE__ */ jsx(Fragment, { children: bandSeries.map((series) => {
    const seriesMetadata = safeGet(seriesMap, series);
    const { color, stroke, opacity } = getBandStyle(
      seriesMetadata,
      series.strokeOnly
    );
    const strokeColor = stroke ?? "";
    return /* @__PURE__ */ jsx(
      AreaCanvas,
      {
        data: series.datapoints,
        canvasWidth: width,
        canvasHeight: height,
        x: x(xScale, datapointsMap),
        y1: y1(yScaleByUnit, seriesMetadata, datapointsMap),
        y0: y0(yScaleByUnit, seriesMetadata, datapointsMap),
        color: () => color,
        strokeColor: () => strokeColor,
        defined: (dataPoint) => definedAccessor(dataPoint, datapointsMap),
        opacity,
        displayScatterPlot,
        strokeWidth: DEFAULT_STROKE_WIDTH_BAND_CHART,
        curve,
        enableDotOverflow: false
      },
      seriesMetadata.id
    );
  }) });
};
BandChartDisplay["displayName"] = "BandChartDisplay";
export {
  BandChartDisplay
};
