// @dynatrace/strato-components/esm/filters/timeframe-selector/components/custom-timeframe-selector/CustomTimeframeSelector.js
import { jsx, jsxs } from "react/jsx-runtime";
import {
  useCallback,
  useEffect,
  useMemo,
  useReducer,
  useRef
} from "react";
import { useIntl } from "react-intl";
import BreakpointsJS from "@dynatrace/strato-design-tokens/breakpoints-js";
import * as customTimeframeSelectorStyles from "./CustomTimeframeSelector.sty.js";
import { Button } from "../../../../buttons/button/Button.js";
import { useId } from "../../../../core/hooks/useId.js";
import { useTimeframeSelectorStore } from "../../../../core/time/timeframeStore/useTimeframeStore.js";
import {
  getTimeBoundaryMessageData,
  validate
} from "../../../../core/time/validation.js";
import { DateTimeRangePickerRoot } from "../../../../forms/date-time-range-picker/DateTimeRangePickerRoot.js";
import { FormFieldMessages } from "../../../../forms/forms-core/components/form-field-messages/FormFieldMessages.js";
import { Label } from "../../../../forms/forms-core/components/label/Label.js";
import { _FormFieldContext as FormFieldContext } from "../../../../forms/forms-core/contexts/FormFieldContext.js";
import { FormFieldMessageContext } from "../../../../forms/forms-core/contexts/FormFieldMessageContext.js";
import { useFormFieldContext } from "../../../../forms/forms-core/hooks/useFormFieldContext.js";
import { useFormFieldMessagesContext } from "../../../../forms/forms-core/hooks/useFormFieldMessagesContext.js";
import {
  FormFieldMessagesAction,
  formFieldMessagesReducer
} from "../../../../forms/forms-core/state/formfield-messages-state-reducer.js";
import { sortTimeframeValues } from "../../../../forms/forms-core/time/helpers.js";
import {
  getFromToValues,
  parseValue
} from "../../../../forms/forms-core/time/timeParser.js";
import { Flex } from "../../../../layouts/flex/Flex.js";
import { useBreakpoint } from "../../../../layouts/hooks/useBreakpoint.js";
import { InputGroup } from "../../../../layouts/input-group/InputGroup.js";
import {
  TIMEFRAMESELECTOR_ERROR_MESSAGES,
  TIMEFRAME_SELECTOR_MESSAGES
} from "../../constants/timeframeSelectorMessages.js";
import { saveRecentTimeframe } from "../../utils/presets-helpers.js";
import { isOutsideRange } from "../../utils/timeframe-selection.js";
const CustomTimeframeSelector = () => {
  const ref = useTimeframeSelectorStore(
    (state2) => state2.dateTimeRangePickerRef
  );
  const intl = useIntl();
  const value = useTimeframeSelectorStore(
    (state2) => state2.dateTimeRangePickerTimeframe
  );
  const onChange = useTimeframeSelectorStore(
    (state2) => state2.setTimeframeSelectorTimeframe
  );
  const min = useTimeframeSelectorStore((state2) => state2.min);
  const max = useTimeframeSelectorStore((state2) => state2.max);
  const precision = useTimeframeSelectorStore((state2) => state2.precision);
  const required = useTimeframeSelectorStore((state2) => state2.required);
  const dateTimeRangePickerError = useTimeframeSelectorStore(
    (state2) => state2.dateTimeRangePickerError
  );
  const formatString = useTimeframeSelectorStore((state2) => state2.formatString);
  const {
    state: { showError }
  } = useFormFieldMessagesContext();
  const rootRef = useRef(null);
  const dateTimeRangePickerId = useId();
  const hintId = useId();
  const labelId = useId();
  const { from, to } = parseValue(value);
  const { required: contextRequired } = useFormFieldContext();
  const callback = useCallback(
    (state2, action) => formFieldMessagesReducer(state2, action),
    []
  );
  const [state, dispatch] = useReducer(callback, {
    customMessages: [],
    formControlMessages: [],
    showError,
    isTouched: showError
  });
  const applyButtonDataDtName = "Apply";
  const { valueFrom, valueTo } = getFromToValues(value);
  const applyButtonDataDtProperties = `valueFrom:${valueFrom},valueTo:${valueTo}`;
  const previousErrorRef = useRef();
  useEffect(() => {
    previousErrorRef.current = validate(value, min, max, {
      formMessageDispatch: dispatch,
      formerError: previousErrorRef.current,
      precision,
      showError,
      checkOrder: true,
      ref,
      expectedFormat: formatString,
      valueMode: "dateTimePicker",
      required: required || contextRequired,
      errorMessages: TIMEFRAMESELECTOR_ERROR_MESSAGES
    });
    if (!required && !contextRequired) {
      if (!from && to?.value === "now()") {
        dispatch({
          type: FormFieldMessagesAction.ADDMESSAGE,
          payload: {
            id: hintId,
            variant: "error",
            message: intl.formatMessage(
              TIMEFRAMESELECTOR_ERROR_MESSAGES.missingFrom
            ),
            messageType: "formcontrol",
            validityStateKey: "customError"
          }
        });
        return;
      }
      if (!to && from?.value === "now()") {
        dispatch({
          type: FormFieldMessagesAction.ADDMESSAGE,
          payload: {
            id: hintId,
            variant: "error",
            message: intl.formatMessage(
              TIMEFRAMESELECTOR_ERROR_MESSAGES.missingTo
            ),
            messageType: "formcontrol",
            validityStateKey: "customError"
          }
        });
        return;
      }
    }
  }, [
    value,
    min,
    max,
    dispatch,
    precision,
    formatString,
    required,
    contextRequired,
    showError,
    ref
  ]);
  const handleApply = useCallback(() => {
    dispatch({
      type: FormFieldMessagesAction.SETISTOUCHED,
      payload: { isTouched: true }
    });
    if (dateTimeRangePickerError) {
      return;
    }
    const currentValue = ref.current?.getCurrentValue();
    const refFrom = currentValue?.from?.value;
    const refTo = currentValue?.to?.value;
    let currentTimeframe = value;
    const onlyFromIsEmpty = !refFrom && !!refTo && refTo !== "now()";
    const onlyToIsEmpty = !refTo && !!refFrom && refFrom !== "now()";
    if (onlyFromIsEmpty || onlyToIsEmpty) {
      const from3 = ref.current?.inputRef.from?.value || "now()";
      const to3 = ref.current?.inputRef.to?.value || "now()";
      const parsedTimeframe = parseValue({ from: from3, to: to3 }, false, precision);
      if (parsedTimeframe.from && parsedTimeframe.to) {
        if (isOutsideRange(
          {
            from: parsedTimeframe.from.value,
            to: parsedTimeframe.to.value
          },
          min?.value,
          max?.value
        )) {
          return;
        }
        const sortedTimeframe = sortTimeframeValues({
          from: parsedTimeframe.from,
          to: parsedTimeframe.to
        });
        if (sortedTimeframe.from && sortedTimeframe.to) {
          currentTimeframe = {
            from: sortedTimeframe.from,
            to: sortedTimeframe.to
          };
        }
      }
    }
    const { from: from2, to: to2 } = currentTimeframe ? parseValue(currentTimeframe, false, precision) : { from: null, to: null };
    if (from2 && to2) {
      const timeframe = { from: from2, to: to2 };
      onChange(timeframe);
      saveRecentTimeframe(timeframe);
      return;
    }
    if (!required && !contextRequired) {
      const hasFromValue = Boolean(refFrom);
      const hasToValue = Boolean(refTo);
      if (!hasFromValue && !hasToValue) {
        onChange(null);
      }
    }
  }, [
    dateTimeRangePickerError,
    ref,
    value,
    precision,
    required,
    contextRequired,
    min,
    max,
    onChange
  ]);
  const handleKeyDown = useCallback(
    (e) => {
      if (e.key !== "Enter" || // Checks if the event target is a child of the DateTimeRangePicker root element ref.
      !ref.current?.element?.contains(e.target)) {
        return;
      }
      ref.current?.validate();
      handleApply();
    },
    [handleApply, ref]
  );
  const isScreenLargerThanMobile = useBreakpoint(
    `(min-width: ${BreakpointsJS.Max.Mobile}px)`,
    true
  );
  useEffect(() => {
    const timeBoundaryMessageData = getTimeBoundaryMessageData(
      {
        from,
        to
      },
      min,
      max,
      precision,
      false
    );
    if (!timeBoundaryMessageData) {
      if (!required && !contextRequired) {
        if (!from && to?.value === "now()") {
          dispatch({
            type: FormFieldMessagesAction.ADDMESSAGE,
            payload: {
              id: hintId,
              variant: "error",
              message: intl.formatMessage(
                TIMEFRAMESELECTOR_ERROR_MESSAGES.missingFrom
              ),
              messageType: "formcontrol",
              validityStateKey: "customError"
            }
          });
          return;
        }
        if (!to && from?.value === "now()") {
          dispatch({
            type: FormFieldMessagesAction.ADDMESSAGE,
            payload: {
              id: hintId,
              variant: "error",
              message: intl.formatMessage(
                TIMEFRAMESELECTOR_ERROR_MESSAGES.missingTo
              ),
              messageType: "formcontrol",
              validityStateKey: "customError"
            }
          });
          return;
        }
      }
      dispatch({
        type: FormFieldMessagesAction.REMOVEMESSAGE,
        payload: {
          id: hintId,
          messageType: "formcontrol"
        }
      });
      return;
    }
    dispatch({
      type: FormFieldMessagesAction.ADDMESSAGE,
      payload: {
        id: hintId,
        variant: "hint",
        message: intl.formatMessage(
          TIMEFRAMESELECTOR_ERROR_MESSAGES[timeBoundaryMessageData.errorKey],
          timeBoundaryMessageData.values
        ),
        messageType: "formcontrol",
        validityStateKey: "customError"
      }
    });
  }, [
    min,
    max,
    dispatch,
    intl,
    precision,
    hintId,
    from,
    to,
    required,
    contextRequired
  ]);
  const messageContextValue = useMemo(
    () => ({
      dispatch,
      state
    }),
    [dispatch, state]
  );
  return /* @__PURE__ */ jsx("div", { ref: rootRef, className: customTimeframeSelectorStyles.overlayRoot, children: /* @__PURE__ */ jsxs(FormFieldMessageContext.Provider, { value: messageContextValue, children: [
    /* @__PURE__ */ jsxs(Flex, { className: customTimeframeSelectorStyles.wrapper, children: [
      /* @__PURE__ */ jsx("div", { onKeyDown: handleKeyDown, children: /* @__PURE__ */ jsx(FormFieldContext.Provider, { value: {}, children: /* @__PURE__ */ jsxs(
        "fieldset",
        {
          className: customTimeframeSelectorStyles.fieldset,
          "aria-labelledby": labelId,
          children: [
            /* @__PURE__ */ jsx(
              Label,
              {
                htmlFor: dateTimeRangePickerId,
                required: required || contextRequired,
                id: labelId,
                children: intl.formatMessage(
                  TIMEFRAME_SELECTOR_MESSAGES.timeframeLabel
                )
              }
            ),
            /* @__PURE__ */ jsxs(InputGroup, { style: { gridAutoColumns: "1fr auto" }, children: [
              /* @__PURE__ */ jsx(
                DateTimeRangePickerRoot,
                {
                  ref,
                  id: dateTimeRangePickerId,
                  required: required || contextRequired,
                  style: { gridTemplateColumns: "1fr auto" },
                  onBlur: (event) => {
                    const hasPresetItemSelected = event.relatedTarget?.getAttribute("role") === "option";
                    const isActiveElementInside = ref.current?.element?.contains(document.activeElement);
                    if (hasPresetItemSelected || isActiveElementInside) {
                      return;
                    }
                    dispatch({
                      type: FormFieldMessagesAction.SETISTOUCHED,
                      payload: { isTouched: true }
                    });
                  }
                }
              ),
              isScreenLargerThanMobile && /* @__PURE__ */ jsx(
                Button,
                {
                  color: "primary",
                  variant: "accent",
                  type: "submit",
                  onClick: handleApply,
                  style: {
                    alignSelf: "end",
                    borderTopLeftRadius: 0,
                    borderBottomLeftRadius: 0
                  },
                  "data-dt-name": applyButtonDataDtName,
                  "data-dt-properties": applyButtonDataDtProperties,
                  children: intl.formatMessage(
                    TIMEFRAME_SELECTOR_MESSAGES.applyButton
                  )
                }
              )
            ] })
          ]
        }
      ) }) }),
      /* @__PURE__ */ jsx(FormFieldMessages, {})
    ] }),
    !isScreenLargerThanMobile && /* @__PURE__ */ jsx(
      Button,
      {
        color: "primary",
        variant: "accent",
        type: "submit",
        style: { alignSelf: "end" },
        onClick: handleApply,
        "data-dt-name": applyButtonDataDtName,
        "data-dt-properties": applyButtonDataDtProperties,
        children: intl.formatMessage(TIMEFRAME_SELECTOR_MESSAGES.applyButton)
      }
    )
  ] }) });
};
CustomTimeframeSelector["displayName"] = "CustomTimeframeSelector";
export {
  CustomTimeframeSelector
};
