// @dynatrace/strato-components/charts/core/components/axes/Axis.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Axis_exports = {};
__export(Axis_exports, {
  Axis: () => Axis
});
module.exports = __toCommonJS(Axis_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_axis = require("@visx/axis");
var import_Axis_css = require("./Axis.sty.js");
var import_constants = require("./constants.js");
var import_chars_font_width_map = require("../text-ellipsis/chars-font-width-map.js");
var import_truncate_string = require("../text-ellipsis/truncate-string.js");
const labelProps = { ...import_constants.axisFontProps, textAnchor: "middle" };
const labelBaseProps = {
  ...import_constants.axisFontBaseProps,
  textAnchor: "middle"
};
const Axis = ({
  width = "100%",
  height = "100%",
  containerSize,
  label,
  size = "small",
  tickLabelProps = (...args) => ({}),
  tickColor,
  ...restProps
}) => {
  const fontProps = {
    ...size === "small" ? import_constants.axisFontProps : import_constants.axisFontBaseProps,
    fill: tickColor ?? import_constants.DEFAULT_AXIS_FONT_COLOR
  };
  const defaultTickLabelProps = (...args) => ({
    ...fontProps,
    ...typeof tickLabelProps === "function" ? tickLabelProps(...args) : tickLabelProps
  });
  const truncatedLabel = label ? (0, import_truncate_string.truncateString)(
    label,
    containerSize,
    size === "small" ? import_chars_font_width_map.charsFontWidth12pixelsSize400Map : import_chars_font_width_map.charsFontWidth14pixelsSize400Map,
    "middle"
  ) : "";
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "svg",
    {
      className: import_Axis_css.axis,
      width,
      height,
      role: restProps.role,
      "aria-label": restProps["aria-label"] ?? void 0,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_axis.Axis,
        {
          label: truncatedLabel,
          labelProps: size === "small" ? labelProps : labelBaseProps,
          tickLabelProps: defaultTickLabelProps,
          tickLength: import_constants.AXIS_TICK_LENGTH,
          hideTicks: true,
          hideAxisLine: true,
          ...restProps
        }
      )
    }
  );
};
Axis["displayName"] = "Axis";
