// @dynatrace/strato-components/charts/core/components/axes/constants.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var constants_exports = {};
__export(constants_exports, {
  AXIS_LABEL_HEIGHT: () => AXIS_LABEL_HEIGHT,
  AXIS_TICK_LENGTH: () => AXIS_TICK_LENGTH,
  BOTTOM_AXIS_LINE_OFFSET: () => BOTTOM_AXIS_LINE_OFFSET,
  DEFAULT_AXIS_FONT_COLOR: () => DEFAULT_AXIS_FONT_COLOR,
  EMOJI_REGEX: () => EMOJI_REGEX,
  MAX_ANNOTATIONS_AREA_HEIGHT: () => MAX_ANNOTATIONS_AREA_HEIGHT,
  MIN_AXIS_TRUNCATION_SIZE: () => MIN_AXIS_TRUNCATION_SIZE,
  MIN_SCALE_STEPS: () => MIN_SCALE_STEPS,
  TICK_ALIGMENT: () => TICK_ALIGMENT,
  TICK_LINE_HEIGHT: () => TICK_LINE_HEIGHT,
  TICK_LINE_ICON_OFFSET: () => TICK_LINE_ICON_OFFSET,
  TOP_AXIS_LABEL_OFFSET: () => TOP_AXIS_LABEL_OFFSET,
  TOP_AXIS_LINE_OFFSET: () => TOP_AXIS_LINE_OFFSET,
  X_AXES_LABEL_OFFSET: () => X_AXES_LABEL_OFFSET,
  X_AXES_TICK_LINE_OFFSET: () => X_AXES_TICK_LINE_OFFSET,
  X_LABEL_AREA: () => X_LABEL_AREA,
  Y_AXES_TICK_LINE_OFFSET: () => Y_AXES_TICK_LINE_OFFSET,
  Y_AXIS_LABEL_OFFSET: () => Y_AXIS_LABEL_OFFSET,
  Y_LABEL_AREA: () => Y_LABEL_AREA,
  axisFontBaseProps: () => axisFontBaseProps,
  axisFontProps: () => axisFontProps
});
module.exports = __toCommonJS(constants_exports);
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
var import_typography = __toESM(require("@dynatrace/strato-design-tokens/typography"));
const AXIS_TICK_LENGTH = 0;
const X_AXES_TICK_LINE_OFFSET = 4;
const Y_AXES_TICK_LINE_OFFSET = 6;
const AXIS_LABEL_HEIGHT = 12;
const MIN_SCALE_STEPS = AXIS_LABEL_HEIGHT;
const FLEX_OFFSET = 4;
const X_AXES_LABEL_OFFSET = X_AXES_TICK_LINE_OFFSET + 6 - FLEX_OFFSET;
const Y_AXIS_LABEL_OFFSET = 6;
const TOP_AXIS_LINE_OFFSET = 18;
const TOP_AXIS_LABEL_OFFSET = 8;
const BOTTOM_AXIS_LINE_OFFSET = 8;
const MIN_AXIS_TRUNCATION_SIZE = 33;
const MAX_ANNOTATIONS_AREA_HEIGHT = 46;
const TICK_LINE_HEIGHT = 14;
const TICK_LINE_ICON_OFFSET = 4;
const Y_LABEL_AREA = AXIS_LABEL_HEIGHT + Y_AXIS_LABEL_OFFSET;
const X_LABEL_AREA = AXIS_LABEL_HEIGHT + X_AXES_LABEL_OFFSET;
const TICK_ALIGMENT = {
  Start: "start",
  Middle: "middle",
  End: "end"
};
const DEFAULT_AXIS_FONT_COLOR = import_colors.default.Text.Neutral.Subdued;
const axisFontProps = {
  fill: DEFAULT_AXIS_FONT_COLOR,
  color: DEFAULT_AXIS_FONT_COLOR,
  fontSize: import_typography.default.Text.Small.Emphasized.Size,
  fontFamily: import_typography.default.Text.Small.Emphasized.Family
};
const axisFontBaseProps = {
  fill: DEFAULT_AXIS_FONT_COLOR,
  color: DEFAULT_AXIS_FONT_COLOR,
  fontSize: import_typography.default.Text.Base.Emphasized.Size,
  fontFamily: import_typography.default.Text.Base.Emphasized.Family
};
const EMOJI_REGEX = /(\p{Emoji_Presentation}|\p{Extended_Pictographic})/u;
