// @dynatrace/strato-components/charts/core/components/chart-states/ChartStateHandler.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ChartStateHandler_exports = {};
__export(ChartStateHandler_exports, {
  ChartStateHandler: () => ChartStateHandler
});
module.exports = __toCommonJS(ChartStateHandler_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = require("clsx");
var import_react_error_boundary = require("react-error-boundary");
var import_ChartState_css = require("./ChartState.sty.js");
var import_EmptyState = require("./EmptyState.js");
var import_ErrorState = require("./ErrorState.js");
var import_useBehavioralTrackingProps = require("../../../../core/hooks/useBehavioralTrackingProps.js");
var import_LoaderOverlay = require("../loader-overlay/LoaderOverlay.js");
const getContent = (props) => {
  const { empty, height, emptyState, errorState, children, isLoading } = props;
  if (empty) {
    return isLoading ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_LoaderOverlay.LoaderOverlay, {}) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_EmptyState.EmptyState, { height, children: emptyState });
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_react_error_boundary.ErrorBoundary,
    {
      FallbackComponent: (0, import_ErrorState.ErrorState)({
        height,
        children: errorState
      }),
      children: [
        children,
        isLoading && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_LoaderOverlay.LoaderOverlay, {})
      ]
    }
  );
};
function ChartStateHandler(props) {
  const {
    height,
    empty,
    emptyState,
    errorState,
    isLoading,
    style,
    className,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    ...rest
  } = props;
  const content = getContent(props);
  const [behavioralTrackingProps, otherProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(rest);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      ...behavioralTrackingProps,
      className: (0, import_clsx.clsx)(import_ChartState_css.container, className),
      style: {
        height,
        filter: isLoading ? "opacity(85%)" : "none",
        ...style
      },
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...otherProps,
      children: content
    }
  );
}
