// @dynatrace/strato-components/charts/core/components/chart-states/ErrorState.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ErrorState_exports = {};
__export(ErrorState_exports, {
  ErrorState: () => ErrorState
});
module.exports = __toCommonJS(ErrorState_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_ChartState_css = require("./ChartState.sty.js");
var import_Flex = require("../../../../layouts/flex/Flex.js");
var import_Text = require("../../../../typography/text/Text.js");
const ErrorStateTextWrapper = ({ height, children }) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
  import_Flex.Flex,
  {
    justifyContent: "center",
    alignItems: "center",
    height,
    width: "100%",
    children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { className: import_ChartState_css.errorText, textStyle: "base", children })
  }
);
const ErrorState = ({ height, children }) => (fallbackProps) => {
  let errorNode;
  if (children && typeof children === "function") {
    errorNode = children(fallbackProps.error.message);
  } else {
    const errorMsg = fallbackProps.error.message ? `: ${fallbackProps.error.message}` : "";
    errorNode = children || /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { className: import_ChartState_css.wrapper, children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.CriticalIcon, {}),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_react_intl.FormattedMessage,
        {
          id: "9SDB/Tz9eGlVpihz",
          defaultMessage: "An error occurred",
          description: "Message to display when there is an error"
        }
      ),
      errorMsg
    ] });
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(ErrorStateTextWrapper, { height, children: errorNode });
};
ErrorState["displayName"] = "ErrorState";
