// @dynatrace/strato-components/tables/DataTable/components/SettingsModal/DataTableSettingsModal.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DataTableSettingsModal_exports = {};
__export(DataTableSettingsModal_exports, {
  DataTableSettingsModal: () => DataTableSettingsModal
});
module.exports = __toCommonJS(DataTableSettingsModal_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_Button = require("../../../../buttons/button/Button.js");
var import_useId = require("../../../../core/hooks/useId.js");
var import_Flex = require("../../../../layouts/flex/Flex.js");
var import_Modal = require("../../../../overlays/modal/Modal.js");
var import_screen_reader_only_css = require("../../../../styles/screen-reader-only.sty.js");
var import_messages = require("../../messages.js");
const DataTableSettingsModal = (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    title,
    dismissible = true,
    show,
    showResetStatus = false,
    children,
    onDismiss,
    onReset,
    onApply,
    controlLabels,
    disableButtons,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    className: consumerClassName,
    style: consumerStyle
  } = props;
  const intl = (0, import_react_intl.useIntl)();
  const statusId = (0, import_useId.useId)();
  const showFooter = onDismiss || onReset || onApply;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Modal.Modal,
    {
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ref: forwardedRef,
      className: consumerClassName,
      style: consumerStyle,
      title,
      size: "small",
      show,
      dismissible,
      onDismiss,
      footer: showFooter && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { justifyContent: "space-between", width: "100%", children: [
        onReset && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_Button.Button,
            {
              onClick: onReset,
              variant: "emphasized",
              disabled: disableButtons?.reset === true,
              children: controlLabels?.reset
            }
          ),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "div",
            {
              id: statusId,
              role: "status",
              "aria-live": "polite",
              "aria-relevant": "all",
              className: import_screen_reader_only_css.screenReaderOnly,
              children: showResetStatus && intl.formatMessage(import_messages.dataTableMessages.resetSuccessful)
            }
          )
        ] }),
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { justifyContent: "flex-end", gap: 8, children: [
          onDismiss && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button, { onClick: onDismiss, variant: "default", children: controlLabels?.cancel }),
          onApply && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_Button.Button,
            {
              onClick: onApply,
              variant: "accent",
              color: "primary",
              disabled: disableButtons?.apply === true,
              children: controlLabels?.apply
            }
          )
        ] })
      ] }),
      children
    }
  );
});
