// @dynatrace/strato-components/content/avatar-group/AvatarGroup.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var AvatarGroup_exports = {};
__export(AvatarGroup_exports, {
  AvatarGroup: () => AvatarGroup
});
module.exports = __toCommonJS(AvatarGroup_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_AvatarGroup_css = require("./AvatarGroup.sty.js");
var import_Item = require("./Item.js");
var import_Button = require("../../buttons/button/Button.js");
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_useRovingFocus = require("../../core/hooks/useRovingFocus.js");
var import_useSafeForwardProps = require("../../core/hooks/useSafeForwardProps.js");
var import_focusRingStatic_css = require("../../core/styles/focusRingStatic.sty.js");
var import_is_slot = require("../../core/utils/is-slot.js");
var import_Flex = require("../../layouts/flex/Flex.js");
var import_Menu = require("../../navigation/menu/Menu.js");
const COMPONENT_NAME = "AvatarGroup";
const MAX_AVATARGROUP_ITEMS = 5;
const RootAvatarGroup = (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const {
      children,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
    const intl = (0, import_react_intl.useIntl)();
    const ariaLabelingProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(restProps);
    const forwardProps = (0, import_useSafeForwardProps.useSafeForwardProps)(restProps, ariaLabelingProps);
    const { "aria-label": ariaLabel, ...remainingAriaLabelingProps } = ariaLabelingProps;
    const { containerRef: rovingContainerRef } = (0, import_useRovingFocus.useRovingFocus)(
      {
        initialIndex: void 0,
        direction: "horizontal",
        focusableChildrenSelector: ".rovingFocusElement"
      }
    );
    const [displayValues, additionalValues] = (0, import_react.useMemo)(() => {
      const childrenArray = import_react.Children.toArray(children);
      let displayValues2;
      const additionalValues2 = [];
      if (childrenArray.length > MAX_AVATARGROUP_ITEMS) {
        displayValues2 = childrenArray.splice(0, MAX_AVATARGROUP_ITEMS);
        for (let i = 0; i < childrenArray.length; i++) {
          const child = childrenArray[i];
          if ((0, import_is_slot.isSlot)(child, import_Item.Item)) {
            additionalValues2.push(
              (0, import_react.cloneElement)(
                /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  import_Item.Item,
                  {
                    abbreviation: child.props.abbreviation,
                    tooltipText: child.props.tooltipText
                  }
                ),
                {
                  ...child.props,
                  key: i,
                  isCollapsed: true
                }
              )
            );
          }
        }
      } else {
        displayValues2 = childrenArray;
      }
      return [displayValues2, additionalValues2];
    }, [children]);
    const rootRef = (0, import_useMergeRefs.useMergeRefs)([forwardedRef, rovingContainerRef]);
    const formatMessageAriaLabel = intl.formatMessage(
      {
        id: "4ER0rfN8lI3HtaFd",
        defaultMessage: "Group of Avatars. {hiddenAvatars} hidden, {totalAvatars} total users",
        description: "Label for the collapsed Avatars."
      },
      {
        hiddenAvatars: additionalValues.length,
        totalAvatars: additionalValues.length + displayValues.length
      }
    );
    return displayValues.length ? /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      import_Flex.Flex,
      {
        gap: 4,
        alignItems: "center",
        ref: rootRef,
        className: (0, import_clsx.default)((0, import_focusRingStatic_css.focusRingStatic)(), consumerClassName),
        style: consumerStyle,
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        "aria-label": ariaLabel ?? formatMessageAriaLabel,
        ...remainingAriaLabelingProps,
        ...behavioralTrackingProps,
        ...forwardProps,
        "data-dt-component": COMPONENT_NAME,
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Flex.Flex, { alignItems: "center", gap: 0, children: displayValues }),
          additionalValues.length > 0 && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Menu.Menu, { children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Trigger, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
              import_Button.Button,
              {
                size: "condensed",
                variant: "emphasized",
                className: (0, import_clsx.default)(
                  "rovingFocusElement",
                  consumerClassName,
                  import_AvatarGroup_css.additionalAvatarWrapperCSS
                ),
                children: [
                  "+",
                  additionalValues.length
                ]
              }
            ) }),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_Menu.Menu.Content,
              {
                "data-dtrum-mask": dataDtrumMask,
                "data-dtrum-allow": dataDtrumAllow,
                children: additionalValues
              }
            )
          ] })
        ]
      }
    ) : null;
  }
);
RootAvatarGroup.displayName = COMPONENT_NAME;
const AvatarGroup = Object.assign(RootAvatarGroup, {
  Item: import_Item.Item
});
