// @dynatrace/strato-components/content/avatar-group/Item.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Item_exports = {};
__export(Item_exports, {
  Item: () => Item
});
module.exports = __toCommonJS(Item_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_AvatarGroup_css = require("./AvatarGroup.sty.js");
var import_Menu = require("../../navigation/menu/Menu.js");
var import_Tooltip = require("../../overlays/tooltip/Tooltip.js");
var import_Avatar = require("../avatar/Avatar.js");
const Item = /* @__PURE__ */ (0, import_react.forwardRef)(
  (props, ref) => {
    const {
      size = "default",
      abbreviation,
      disabled = false,
      tooltipText,
      as,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...remainingProps
    } = props;
    const isCollapsed = remainingProps.isCollapsed;
    return !isCollapsed ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "div",
      {
        className: (0, import_clsx.default)(consumerClassName, import_AvatarGroup_css.avatarGroupItemWrapperCSS),
        style: consumerStyle,
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_Tooltip.Tooltip,
          {
            text: tooltipText,
            fallbackPlacements: [
              "top",
              "bottom",
              "top-start",
              "top-end",
              "bottom-start",
              "bottom-end",
              "right",
              "left",
              "right-start",
              "right-end",
              "left-start",
              "left-end"
            ],
            "data-dtrum-mask": dataDtrumMask,
            "data-dtrum-allow": dataDtrumAllow,
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_Avatar.Avatar,
              {
                abbreviation,
                ref,
                size,
                as,
                className: "rovingFocusElement",
                disabled,
                ...remainingProps
              }
            )
          }
        )
      }
    ) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Menu.Menu.Item,
      {
        "data-testid": dataTestId,
        onSelect: (() => null),
        className: (0, import_clsx.default)("rovingFocusElement", consumerClassName),
        as,
        disabled,
        style: consumerStyle,
        ...remainingProps,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Avatar.Avatar, { abbreviation, size: "small", disabled, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Avatar.Avatar.Label, { children: tooltipText }) })
      }
    );
  }
);
Item.displayName = "AvatarGroup.Item";
