// @dynatrace/strato-components/charts/histogram/components/axes/HistogramValueBottomAxisArea/HistogramValueBottomAxisArea.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var HistogramValueBottomAxisArea_exports = {};
__export(HistogramValueBottomAxisArea_exports, {
  HistogramValueBottomAxisArea: () => HistogramValueBottomAxisArea
});
module.exports = __toCommonJS(HistogramValueBottomAxisArea_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("@use-gesture/react");
var import_clsx = __toESM(require("clsx"));
var import_lodash_es = require("lodash");
var import_react2 = require("react");
var import_HistogramPanningControls = require("../HistogramPanningControls.js");
var import_HistogramValueBottomAxisArea_css = require("./HistogramValueBottomAxisArea.sty.js");
var import_ValueBottomAxis = require("../../../../core/components/axes/components/ValueBottomAxis.js");
var import_ChartGridRoot = require("../../../../core/components/chart-grid/ChartGridRoot.js");
var import_constants = require("../../../../core/components/zoom/constants.js");
var import_ZoomCursors_css = require("../../../../core/components/zoom/ZoomCursors.sty.js");
var import_histogram_grid_area_rules = require("../../../constants/histogram-grid-area-rules.js");
var import_selectors = require("../../../hooks/store/selectors.js");
var import_useHistogramInteractions = require("../../../hooks/useHistogramInteractions.js");
var import_useHistogramTickAxis = require("../../../hooks/useHistogramTickAxis.js");
var import_useHistogramXScale = require("../../../hooks/useHistogramXScale.js");
var import_useXAxisConfig = require("../../../hooks/useXAxisConfig.js");
var import_useApplyPanDelta = require("../../overlays/hooks/useApplyPanDelta.js");
const HistogramValueBottomAxisArea = () => {
  const axisRef = (0, import_react2.useRef)(null);
  const { label } = (0, import_useXAxisConfig.useXAxisConfig)();
  const ticks = (0, import_useHistogramTickAxis.useHistogramTickAxis)();
  const { pan } = (0, import_useHistogramInteractions.useHistogramInteractions)();
  const { zoomOutEnabled } = (0, import_selectors.useZoomState)();
  const { xScale } = (0, import_useHistogramXScale.useHistogramXScale)();
  const isPanEnabled = pan?.enabled && zoomOutEnabled;
  const applyPan = (0, import_useApplyPanDelta.useApplyPanDelta)();
  (0, import_react.useGesture)(
    {
      onDrag: (e) => !import_constants.KEYBOARD_HANDLERS.includes(e.type) && applyPan(-e.delta[0])
    },
    {
      target: axisRef,
      eventOptions: { passive: false },
      drag: { filterTaps: true, pointer: { buttons: [1, 4] } },
      enabled: isPanEnabled
    }
  );
  const bottomTicks = ticks?.bottom || { ticks: [], formattedTicks: [] };
  if ((0, import_lodash_es.isEmpty)(bottomTicks.ticks) || !xScale) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ChartGridRoot.ChartGrid.GridArea, { gridArea: import_histogram_grid_area_rules.AREAS.bottomAxis, children: ({ availableWidth, availableHeight }) => {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      "div",
      {
        className: (0, import_clsx.default)((0, import_ZoomCursors_css.cursorStyle)({ isPanEnabled }), import_HistogramValueBottomAxisArea_css.axisContainerStyle),
        ref: axisRef,
        "data-testid": "histogram-bottom-axis",
        children: [
          isPanEnabled && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_HistogramPanningControls.HistogramPanningControls, { onClickPan: (d) => applyPan(-d) }),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_ValueBottomAxis.ValueBottomAxis,
            {
              ticks: bottomTicks.ticks,
              formattedTicks: bottomTicks.formattedTicks,
              containerHeight: availableHeight,
              containerWidth: availableWidth,
              scale: xScale,
              label
            }
          )
        ]
      }
    );
  } });
};
HistogramValueBottomAxisArea["displayName"] = "HistogramValueBottomAxisArea";
