// @dynatrace/strato-components/esm/filters/segment-selector/hooks/useSegments.js
import { useMutation } from "@tanstack/react-query";
import { useCallback, useContext, useEffect, useRef } from "react";
import {
  addRecentSegmentsPreset
} from "@dynatrace-sdk/segments-presets";
import { useStoreSelector } from "../../../core/store/selector.js";
import { isDev } from "../../../core/utils/is-node-environment.js";
import {
  handleFailedOptimisticMutation,
  handleSettledOptimisticMutation,
  runOptimisticMutation
} from "../../helper/queryHelper.js";
import {
  useSegmentsCache,
  segmentsClient
} from "../contexts/SegmentsCache/SegmentsCache.js";
import { SegmentsContext } from "../contexts/SegmentsStore/SegmentsContext.js";
import { UISegmentsDispatchContext } from "../contexts/SegmentsUI/SegmentsUIDispatchContext.js";
import { areUiSegmentsValid } from "../contexts/SegmentsUI/segmentValidationHelper.js";
import { isSelectedPreset } from "../parts/segmentsPresets/isSelectedPreset.js";
function updateCache(segments, old) {
  if (!old) {
    return { recentPresets: [], pinnedPresets: [] };
  }
  const recentPresets = old.recentPresets;
  const recentPresetIndex = recentPresets.findIndex(
    (preset) => isSelectedPreset(segments, preset.segments)
  );
  if (recentPresetIndex > 0) {
    const oldPreset = old.recentPresets[recentPresetIndex];
    recentPresets.splice(recentPresetIndex, 1);
    recentPresets.unshift(oldPreset);
  }
  if (recentPresetIndex === -1) {
    const tempId = crypto.randomUUID();
    recentPresets.unshift({ id: tempId, segments });
    recentPresets.splice(10);
  }
  return {
    ...old,
    recentPresets
  };
}
const useSegments = () => {
  const store = useContext(SegmentsContext);
  const storeRef = useRef(store);
  const segments = useStoreSelector(
    SegmentsContext,
    (state) => state
  );
  const segmentsRef = useRef(segments);
  const { getSegment, getAllSegments } = useSegmentsCache();
  const { dispatch } = useContext(UISegmentsDispatchContext);
  const getAllSegmentsRef = useRef(getAllSegments);
  const getSegmentRef = useRef(getSegment);
  const { mutate: addToRecentPresetsMutationFn } = useMutation(
    {
      mutationKey: ["add-recent-segment-preset"],
      mutationFn: (segments2) => addRecentSegmentsPreset(segments2),
      onMutate: async () => runOptimisticMutation(
        "getSegmentsPresets",
        (prevState) => updateCache(segments, prevState)
      ),
      onError: (error, _, context) => handleFailedOptimisticMutation("getSegmentsPresets", error, context),
      onSettled: () => handleSettledOptimisticMutation(
        ["getSegmentsPresets"],
        ["add-recent-segment-preset"]
      )
    },
    segmentsClient.queryClient
  );
  useEffect(() => {
    segmentsRef.current = segments;
  }, [segments]);
  useEffect(() => {
    storeRef.current = store;
  }, [store]);
  useEffect(() => {
    getAllSegmentsRef.current = getAllSegments;
  }, [getAllSegments]);
  useEffect(() => {
    getSegmentRef.current = getSegment;
  }, [getSegment]);
  const addSegment = useCallback(
    async (segment) => {
      const isAlreadyInStore = segmentsRef.current.find(
        (storeSegment) => storeSegment.id === segment.id
      );
      if (isAlreadyInStore) {
        return false;
      }
      try {
        const safeSegment = await segmentsClient.queryClient.fetchQuery({
          queryKey: ["getSegment", segment.id],
          queryFn: () => getSegmentRef.current(segment.id)
        });
        if (safeSegment) {
          storeRef.current.setState((state) => {
            const newState = [...state];
            newState.push({ id: segment.id, variables: segment.variables });
            addToRecentPresetsMutationFn(newState);
            return newState;
          });
        }
      } catch (e) {
        if (isDev()) {
          console.warn(
            "SegmentSelector: The segment is not in the segments manager.",
            e
          );
        }
        dispatch({ type: "remove", payload: { id: segment.id } });
        return false;
      }
    },
    [dispatch, addToRecentPresetsMutationFn, getSegmentRef]
  );
  const removeSegment = useCallback((id) => {
    if (!id) {
      return;
    }
    storeRef.current.setState(
      (state) => state.filter((storeSegments) => id !== storeSegments.id)
    );
  }, []);
  const removeAllSegments = useCallback(
    () => storeRef.current.setState([]),
    []
  );
  const setSegments = useCallback(
    async (segments2, force) => {
      if (force) {
        storeRef.current.setState(segments2);
        addToRecentPresetsMutationFn(segments2);
        return;
      }
      if (segmentsClient.queryClient.getQueryData([
        "getAllSegments"
      ]) === void 0) {
        try {
          await segmentsClient.queryClient.fetchQuery({
            queryKey: ["getAllSegments"],
            queryFn: getAllSegmentsRef.current
          });
        } catch (error) {
          if (isDev()) {
            console.warn(
              "SegmentSelector: Couldn't set segments due to a failed fetch.",
              error
            );
          }
          if (segmentsClient.hasAccessRights) {
            console.error(
              "SegmentSelector: Couldn't set segments due to a failed fetch.",
              { cause: error }
            );
            return;
          }
        }
      }
      if (!segmentsClient.hasAccessRights) {
        storeRef.current.setState(segments2);
        return;
      }
      const isValid = await areUiSegmentsValid(segments2);
      if (!isValid) {
        console.error(
          "SegmentSelector: Couldn't set segments due to an invalid selection."
        );
        return;
      }
      storeRef.current.setState(segments2);
      addToRecentPresetsMutationFn(segments2);
    },
    [addToRecentPresetsMutationFn]
  );
  return {
    segments,
    addSegment,
    removeSegment,
    removeAllSegments,
    setSegments
  };
};
export {
  useSegments
};
