// @dynatrace/strato-components/esm/filters/segment-selector/hooks/useTriggerProps.js
import { useQuery } from "@tanstack/react-query";
import { useMemo } from "react";
import { useIntl } from "react-intl";
import { useSegments } from "./useSegments.js";
import { useAriaLabelingProps as _useAriaLabelingProps } from "../../../core/hooks/useAriaLabelingProps.js";
import {
  segmentsClient,
  useSegmentsCache
} from "../contexts/SegmentsCache/SegmentsCache.js";
import { SEGMENTSELECTOR_MESSAGES } from "../segmentMessages.js";
import { getReadableValue } from "../utils.js";
function useTriggerProps(variant, remainingProps) {
  const intl = useIntl();
  const { segments } = useSegments();
  const { getSegment } = useSegmentsCache();
  const safeId = segments[0] ? segments[0].id : void 0;
  const shouldFetchText = variant !== "compact" && segments.length === 1 && safeId !== void 0;
  const { data, isLoading } = useQuery(
    {
      queryKey: ["getSegment", safeId, shouldFetchText],
      queryFn: () => getSegment(safeId),
      enabled: shouldFetchText
    },
    segmentsClient.queryClient
  );
  const text = useMemo(() => {
    if (segments.length === 0 || !safeId) {
      return null;
    }
    if (variant === "compact") {
      return `${segments.length}`;
    }
    if (segments.length > 1) {
      return intl.formatMessage(
        {
          defaultMessage: "{length} segments",
          id: "7vkTbyAZNeY/qa9p",
          description: "Text for SegmentSelector Trigger when multiple segments are selected."
        },
        {
          length: segments.length
        }
      );
    }
    if (!data) {
      return `${safeId}`;
    }
    const amountOfVariables = segments[0]?.variables?.[0]?.values?.length ?? 0;
    if (amountOfVariables === 1) {
      return `${data.segment.name}: ${getReadableValue(
        segments[0].variables?.[0].values[0],
        intl
      )}`;
    }
    if (amountOfVariables > 1) {
      return `${data.segment.name} (${amountOfVariables})`;
    }
    return `${data.segment.name}`;
  }, [data, intl, safeId, segments, variant]);
  const ariaLabelingProps = _useAriaLabelingProps(remainingProps);
  const ariaLabel = useMemo(() => {
    const ariaLabelProp = ariaLabelingProps["aria-label"];
    if (ariaLabelProp === void 0 && ariaLabelingProps["aria-labelledby"] === void 0) {
      return text ? void 0 : intl.formatMessage(SEGMENTSELECTOR_MESSAGES.defaultAriaLabel);
    }
    const combinedAriaLabel = text ? ariaLabelProp?.concat(", ", text) : ariaLabelProp;
    return combinedAriaLabel;
  }, [ariaLabelingProps, intl, text]);
  const ariaProps = {
    ...ariaLabelingProps,
    "aria-label": ariaLabel,
    "aria-disabled": remainingProps["aria-disabled"]
  };
  return { isLoading: shouldFetchText ? isLoading : false, text, ariaProps };
}
export {
  useTriggerProps
};
