// @dynatrace/strato-components/esm/charts/xy-chart/components/annotations/utils/get-indicators-from-annotations.js
import {
  buildIndicator,
  isAlwaysHidden,
  isAlwaysShown
} from "../../../../core/components/annotations/components/indicators/utils/indicator-visibility-helpers.js";
import { DEFAULT_MARKER_COLOR } from "../../../../core/components/annotations/constants.js";
import { getMinMaxAsNumber } from "../../../../core/components/annotations/utils/get-min-max-as-number.js";
import { populateMissingMarkerProps } from "../../../../core/components/annotations/utils/populate-missing-marker-props.js";
import { populateMissingTrackProps } from "../../../../core/components/annotations/utils/populate-missing-track-props.js";
import { serializeMarkerRange } from "../../../../core/components/annotations/utils/serialize-marker-range.js";
import { getDatumValue } from "../../../processing-strategies/utils/access-datapoint/get-datapoints.js";
import { evaluateOptionalAttribute } from "../../../utils/evaluate-with-data.js";
const serializeAnnotationsMarkerFromXYChartData = (data, markerConfig) => {
  const response = {
    ...data,
    start: getDatumValue(data, markerConfig.startAccessor),
    end: getDatumValue(data, markerConfig.endAccessor),
    symbol: evaluateOptionalAttribute(markerConfig.symbol, data, void 0),
    hidden: evaluateOptionalAttribute(markerConfig.hidden, data, false),
    priority: evaluateOptionalAttribute(markerConfig.priority, data, 0),
    color: markerConfig.colorAccessor ? getDatumValue(data, markerConfig.colorAccessor) : void 0,
    title: markerConfig.titleAccessor ? getDatumValue(data, markerConfig.titleAccessor) : void 0,
    description: markerConfig.descriptionAccessor ? getDatumValue(data, markerConfig.descriptionAccessor) : void 0,
    indicatorsDisplay: evaluateOptionalAttribute(
      markerConfig.indicatorsDisplay,
      data,
      void 0
    )
  };
  return response;
};
const serializeAnnotationsTrackFromXYChartData = (data, trackConfig) => {
  return {
    hidden: evaluateOptionalAttribute(trackConfig.hidden, data, false),
    symbol: evaluateOptionalAttribute(trackConfig.symbol, data, void 0),
    color: trackConfig.colorAccessor ? getDatumValue(data, trackConfig.colorAccessor) : void 0,
    label: trackConfig.labelAccessor ? getDatumValue(data, trackConfig.labelAccessor) : void 0,
    indicatorsDisplay: evaluateOptionalAttribute(
      trackConfig.indicatorsDisplay,
      data,
      "auto"
    ),
    markers: []
  };
};
const getIndicatorsFromAnnotations = (data, tracksConfig, markersConfig, selection) => {
  const alwaysVisible = [];
  const selectable = /* @__PURE__ */ new Map();
  const trackIdTroughAccesor = tracksConfig.trackIdAccessor;
  const hiddenTracks = /* @__PURE__ */ new Set();
  const annotationsByTrack = data.reduce(
    (trackMap, datum) => {
      const trackId = getDatumValue(datum, trackIdTroughAccesor);
      const annotationsTrack = trackMap.has(
        trackId
      ) ? trackMap.get(trackId) : populateMissingTrackProps(
        serializeAnnotationsTrackFromXYChartData(datum, tracksConfig)
      );
      if (annotationsTrack?.hidden) {
        hiddenTracks.add(trackId);
      }
      const marker = serializeAnnotationsMarkerFromXYChartData(
        datum,
        markersConfig
      );
      const populatedMarker = populateMissingMarkerProps(marker, annotationsTrack);
      annotationsTrack.markers.push(populatedMarker);
      trackMap.set(trackId, annotationsTrack);
      const { color: trackColor, indicatorsDisplay: trackIndicatorDisplay } = annotationsTrack;
      const {
        color: markerColor,
        start: markerStartRaw,
        end: markerEndRaw,
        indicatorsDisplay: markerIndicatorDisplay,
        priority
      } = populatedMarker;
      const { min: markerStart, max: markerEnd } = getMinMaxAsNumber(
        markerStartRaw,
        markerEndRaw
      );
      if (!isAlwaysHidden(markerIndicatorDisplay, trackIndicatorDisplay)) {
        const trackIndex = Array.from(trackMap.keys()).findIndex(
          (name) => name === trackId
        );
        const markerId = serializeMarkerRange(
          { ...populatedMarker, start: markerStart, end: markerEnd },
          trackIndex
        );
        const color = markerColor ?? trackColor ?? DEFAULT_MARKER_COLOR;
        const indicator = buildIndicator(
          color,
          selection,
          markerId,
          markerStart,
          markerEnd,
          priority
        );
        if (isAlwaysShown(markerIndicatorDisplay, trackIndicatorDisplay)) {
          alwaysVisible.push(indicator);
        } else {
          selectable.set(markerId, indicator);
        }
      }
      return trackMap;
    },
    /* @__PURE__ */ new Map()
  );
  return {
    indicators: { alwaysVisible, selectable },
    annotationsByTrack,
    hiddenTracks
  };
};
export {
  getIndicatorsFromAnnotations
};
