// @dynatrace/strato-components/charts/core/components/tooltip/shared/styles/SeriesBasedContent.sty.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SeriesBasedContent_css_exports = {};
__export(SeriesBasedContent_css_exports, {
  detailViewContent: () => detailViewContent,
  detailViewItem: () => detailViewItem,
  multipleDimensionContainer: () => multipleDimensionContainer,
  multipleDimensionItem: () => multipleDimensionItem,
  singleDimensionContainer: () => singleDimensionContainer,
  valueContainer: () => valueContainer,
  valueItem: () => valueItem
});
module.exports = __toCommonJS(SeriesBasedContent_css_exports);
var import_textStyle_css_ts_vanilla = require("../../../../../../styles/textStyle.css");
var import_SeriesBasedContent_css_ts_vanilla = require("./SeriesBasedContent.css");
var import_createRuntimeFn = require("@vanilla-extract/recipes/createRuntimeFn");
var detailViewContent = "_dacoys0-3-6-0";
var detailViewItem = "_dacoys1-3-6-0";
var multipleDimensionContainer = (0, import_createRuntimeFn.createRuntimeFn)({ defaultClassName: "_dacoys5-3-6-0", variantClassNames: { highlighted: { true: "_dacoys6-3-6-0", false: "_dacoys7-3-6-0" } }, defaultVariants: {}, compoundVariants: [] });
var multipleDimensionItem = "_dacoys8-3-6-0";
var singleDimensionContainer = (0, import_createRuntimeFn.createRuntimeFn)({ defaultClassName: "_dacoys2-3-6-0", variantClassNames: { highlighted: { true: "_dacoys3-3-6-0", false: "_dacoys4-3-6-0" } }, defaultVariants: {}, compoundVariants: [] });
var valueContainer = "_dacoys9-3-6-0";
var valueItem = "_dacoysa-3-6-0 _da9a8v0-3-6-0 _da9a8v2-3-6-0 _da9a8v6-3-6-0 _da9a8ve-3-6-0";
