// @dynatrace/strato-components/charts/core/components/legend/range/components/RangeLegendItemLabel/RangeLegendItemLabel.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var RangeLegendItemLabel_exports = {};
__export(RangeLegendItemLabel_exports, {
  RangeLegendItemLabel: () => RangeLegendItemLabel
});
module.exports = __toCommonJS(RangeLegendItemLabel_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_RangeLegendItemLabel_css = require("./RangeLegendItemLabel.sty.js");
var import_constants = require("../../../../../constants.js");
var import_constants2 = require("../../constants.js");
var import_RangeLegendOrientationContext = require("../../context/RangeLegendOrientationContext.js");
var import_selectors = require("../../hooks/store/selectors.js");
var import_use_legend_intermediate_label_collision_detector = require("../../hooks/use-legend-intermediate-label-collision-detector.js");
const RangeLegendItemLabel = (props) => {
  const { alignment, endPosition, labelId, startPosition, text } = props;
  const { orientation } = (0, import_RangeLegendOrientationContext.useOrientation)();
  const isVertical = orientation === import_constants.ORIENTATION.Vertical;
  const isHorizontal = orientation === import_constants.ORIENTATION.Horizontal;
  const hoveredMarkerId = (0, import_selectors.useRangeLegendHoveredMarkerId)();
  const isLabelHovered = labelId === hoveredMarkerId;
  const {
    text: label,
    x,
    y
  } = (0, import_use_legend_intermediate_label_collision_detector.useLegendIntermediateLabelCollisionDetector)(
    text,
    startPosition,
    endPosition,
    alignment,
    !isLabelHovered
  );
  if ((0, import_lodash_es.isUndefined)(x) || (0, import_lodash_es.isUndefined)(y) || !isLabelHovered) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "text",
    {
      x: x + (isHorizontal ? 0 : import_constants2.LABEL_HOVERED_PADDING),
      y: y + (isVertical ? 0 : import_constants2.LABEL_HOVERED_PADDING),
      fontSize: import_constants2.DEFAULT_LABEL_FONT_SIZE,
      fontWeight: import_constants2.HOVERED_LABEL_FONT_WEIGHT,
      fill: import_constants2.LABEL_FONT_COLOR_HOVERED,
      className: import_RangeLegendItemLabel_css.legendLabel,
      role: "listitem",
      "aria-label": `Label ${text}`,
      children: label
    }
  );
};
