// @dynatrace/strato-components/filters/filter-field/utils/conversion/lezer-tree-conversion/node-type-mapping.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var node_type_mapping_exports = {};
__export(node_type_mapping_exports, {
  nodeTypeOverrides: () => nodeTypeOverrides
});
module.exports = __toCommonJS(node_type_mapping_exports);
const nodeTypeOverrides = {
  FilterGroup: "Group",
  FilterStatement: "Statement",
  FilterKey: "Key",
  FilterSimpleValue: "String",
  FilterIncludesValue: "List",
  Identifier: "String",
  "\u26A0": "Error",
  InvalidComparisonOperator: "Error",
  InvalidInclusionValue: "Error",
  InvalidIdentifier: "Error"
};
