// @dynatrace/strato-components/charts/core/components/overlays/constants.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var constants_exports = {};
__export(constants_exports, {
  BRUSH_HOVER_FILL: () => BRUSH_HOVER_FILL,
  CHART_OVERLAYS_BORDER_WIDTH: () => CHART_OVERLAYS_BORDER_WIDTH,
  KNOB_HEIGHT: () => KNOB_HEIGHT,
  KNOB_WIDTH: () => KNOB_WIDTH,
  SELECTED_AREA_RADIUS: () => SELECTED_AREA_RADIUS,
  SELECTION_AREA_STYLES: () => SELECTION_AREA_STYLES,
  ZOOM_AREA_STYLES: () => ZOOM_AREA_STYLES
});
module.exports = __toCommonJS(constants_exports);
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
const CHART_OVERLAYS_BORDER_WIDTH = 1;
const BRUSH_HOVER_FILL = import_colors.default.Background.Field.Primary.EmphasizedHover;
const SELECTED_AREA_RADIUS = 4;
const SELECTION_AREA_STYLES = {
  fill: import_colors.default.Background.Field.Primary.Accent,
  stroke: "transparent",
  opacity: 0.2
};
const ZOOM_AREA_STYLES = {
  fill: import_colors.default.Background.Field.Neutral.Accent,
  stroke: "transparent",
  opacity: 0.1
};
const KNOB_HEIGHT = 40;
const KNOB_WIDTH = 5;
