// @dynatrace/strato-components/charts/core/components/legend/categorical/components/ChartLegendItem/ChartLegendItem.sty.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ChartLegendItem_css_exports = {};
__export(ChartLegendItem_css_exports, {
  ItemFontStyles: () => ItemFontStyles,
  baseLegendButton: () => baseLegendButton,
  legendItem: () => legendItem,
  legendItemButton: () => legendItemButton,
  legendNameButton: () => legendNameButton,
  legendNameButtonOverride: () => legendNameButtonOverride,
  legendNameWrapper: () => legendNameWrapper
});
module.exports = __toCommonJS(ChartLegendItem_css_exports);
var import_focusRingStatic_css_ts_vanilla = require("../../../../../../../core/styles/focusRingStatic.css");
var import_ChartLegendItem_css_ts_vanilla = require("./ChartLegendItem.css");
var import_createRuntimeFn = require("@vanilla-extract/recipes/createRuntimeFn");
var ItemFontStyles = { fontWeight: "var(--dt-typography-text-small-emphasized-weight, 500)", fontSize: "var(--dt-typography-text-small-emphasized-size, 12px)", fontFamily: "var(--dt-typography-text-small-default-family, DynatraceFlow,Roboto,Helvetica,sans-serif)", lineHeight: "var(--dt-typography-text-small-emphasized-line-height, 1.333333333)" };
var baseLegendButton = "_mt8v6g0-3-6-0 _z0fq9b4-3-6-0 _z0fq9bd-3-6-0 _z0fq9b6-3-6-0 _z0fq9b9-3-6-0";
var legendItem = "_mt8v6g9-3-6-0";
var legendItemButton = (0, import_createRuntimeFn.createRuntimeFn)({ defaultClassName: "_mt8v6g1-3-6-0", variantClassNames: { isInteractive: { true: "_mt8v6g2-3-6-0", false: "_mt8v6g3-3-6-0" } }, defaultVariants: {}, compoundVariants: [] });
var legendNameButton = (0, import_createRuntimeFn.createRuntimeFn)({ defaultClassName: "_mt8v6g4-3-6-0", variantClassNames: { isInteractive: { true: "_mt8v6g5-3-6-0", false: "_mt8v6g6-3-6-0" }, isSubdued: { true: "_mt8v6g7-3-6-0", false: "_mt8v6g8-3-6-0" } }, defaultVariants: {}, compoundVariants: [] });
var legendNameButtonOverride = "_mt8v6gb-3-6-0";
var legendNameWrapper = "_mt8v6ga-3-6-0";
