// @dynatrace/strato-components/charts/core/components/legend/categorical/components/ChartLegendItem/ChartLegendItem.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ChartLegendItem_exports = {};
__export(ChartLegendItem_exports, {
  ChartLegendItem: () => ChartLegendItem
});
module.exports = __toCommonJS(ChartLegendItem_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_ChartLegendItem_css = require("./ChartLegendItem.sty.js");
var import_Tooltip = require("../../../../../../../overlays/tooltip/Tooltip.js");
var import_useLegend = require("../../hooks/useLegend.js");
var import_useLegendItemActions = require("../../hooks/useLegendItemActions.js");
var import_useLegendItemInteractions = require("../../hooks/useLegendItemInteractions.js");
var import_useLegendItemState = require("../../hooks/useLegendItemState.js");
var import_messages = require("../../messages.js");
var import_ChartLegendIcon = require("../ChartLegendIcon/ChartLegendIcon.js");
var import_ChartLegendItemLabel = require("../ChartLegendItemLabel/ChartLegendItemLabel.js");
var import_ChartLegendSeriesActions = require("../ChartLegendSeriesActions/ChartLegendSeriesActions.js");
const ChartLegendItem = (props) => {
  const { item, numberOfItems } = props;
  const {
    hiddenItems,
    onLegendIconClick,
    onItemHover,
    onLegendNameClick,
    interactionsDisabled,
    onSeriesAction: actions
  } = (0, import_useLegend.useLegend)();
  const {
    refIconButton,
    refNameButton,
    open,
    updateOpenState,
    updateHoveredState,
    updateVisible,
    visible,
    isLastItem,
    shouldShowActionsButton,
    isIconHovered
  } = (0, import_useLegendItemState.useLegendItemState)(
    item,
    interactionsDisabled,
    hiddenItems,
    numberOfItems
  );
  const {
    handleLabelMouseEnter,
    handleIconMouseEnter,
    handleMouseLeave,
    onNameButtonClick,
    onNameButtonKeyDown,
    onIconButtonClick
  } = (0, import_useLegendItemInteractions.useLegendItemInteractions)(
    item,
    {
      onItemHover,
      onLegendIconClick,
      onLegendNameClick,
      updateHoveredState,
      updateVisible,
      visible
    },
    interactionsDisabled
  );
  const onLegendItemAction = () => {
    onItemHover(void 0);
    updateHoveredState(false);
  };
  const { intl, defaultActions } = (0, import_useLegendItemActions.useLegendItemActions)(
    item,
    onLegendItemAction
  );
  const iconButtonClickHandler = !isLastItem ? onIconButtonClick : void 0;
  const disabled = interactionsDisabled || isLastItem;
  const memoizedActions = (0, import_react.useMemo)(() => {
    if (open && actions) {
      return actions(item);
    }
  }, [open]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    "div",
    {
      onMouseLeave: handleMouseLeave,
      "data-testid": "chart-legend-item",
      className: import_ChartLegendItem_css.legendItem,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_Tooltip.Tooltip,
          {
            "data-testid": disabled ? "chart-legend-item-tooltip-disabled" : "chart-legend-item-tooltip",
            disabled,
            text: visible ? intl.formatMessage(import_messages.LEGEND_MESSAGES.legendHideSeries) : intl.formatMessage(import_messages.LEGEND_MESSAGES.legendShowSeries),
            placement: "left",
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_ChartLegendIcon.ChartLegendIcon,
              {
                visible,
                hover: isIconHovered,
                disabled: isLastItem,
                color: item.color.Default,
                icon: item.icon,
                buttonRef: refIconButton,
                onMouseEnter: handleIconMouseEnter,
                onButtonClick: iconButtonClickHandler
              }
            )
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
          "div",
          {
            onFocus: handleLabelMouseEnter,
            "data-testid": "chart-legend-name-wrapper",
            className: import_ChartLegendItem_css.legendNameWrapper,
            children: [
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_ChartLegendItemLabel.ChartLegendItemLabel,
                {
                  item,
                  visible,
                  buttonRef: refNameButton,
                  onButtonClick: onNameButtonClick,
                  onButtonKeydown: onNameButtonKeyDown,
                  onMouseEnter: handleLabelMouseEnter,
                  onMouseLeave: handleMouseLeave
                }
              ),
              actions && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_ChartLegendSeriesActions.ChartLegendSeriesActions,
                {
                  disabled: interactionsDisabled,
                  visible: shouldShowActionsButton,
                  onOpenChange: updateOpenState,
                  isOpen: open,
                  onFocus: handleLabelMouseEnter,
                  onMouseEnter: handleLabelMouseEnter,
                  onMouseLeave: handleMouseLeave,
                  defaultActions,
                  actions: memoizedActions,
                  onBlur: handleMouseLeave
                }
              )
            ]
          }
        )
      ]
    }
  );
};
ChartLegendItem["displayName"] = "ChartLegendItem";
