// @dynatrace/strato-components/charts/timeseries/providers/TimeseriesDataProvider.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TimeseriesDataProvider_exports = {};
__export(TimeseriesDataProvider_exports, {
  TimeseriesDataProvider: () => TimeseriesDataProvider,
  useBuildInitialMetadata: () => useBuildInitialMetadata
});
module.exports = __toCommonJS(TimeseriesDataProvider_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_useColorRule = require("../../core/color-rule/useColorRule.js");
var import_get_color_palette = require("../../core/utils/get-color-palette.js");
var import_TimeseriesData_context = require("../context/TimeseriesData.context.js");
var import_useColorPalette = require("../hooks/useColorPalette.js");
var import_useDataByYAxis = require("../hooks/useDataByYAxis.js");
var import_useGapPolicy = require("../hooks/useGapPolicy.js");
var import_usePointsDisplay = require("../hooks/usePointsDisplay.js");
var import_timeseries2 = require("../utils/timeseries.js");
function getAllUnits(data) {
  return (0, import_lodash_es.uniq)(data.shapePriority.flatMap((v) => data[v]?.map((s) => s.unit)));
}
function useBuildInitialMetadata(data, valueRepresentation, mergedData) {
  const colorPalette = (0, import_useColorPalette.useColorPalette)();
  const gapPolicy = (0, import_useGapPolicy.useGapPolicy)();
  const pointsDisplay = (0, import_usePointsDisplay.usePointsDisplay)();
  const seriesMap = /* @__PURE__ */ new WeakMap();
  const allUnits = getAllUnits(data);
  const colorRules = (0, import_useColorRule.useColorRule)(mergedData, "timeseries");
  let startingIndex = 0;
  const valueRepPerVariant = /* @__PURE__ */ new Map();
  for (const variant of data.shapePriority) {
    const variantSeries = data[variant];
    for (let i = 0; i < variantSeries.length; i++) {
      const series = variantSeries[i];
      const id = crypto.randomUUID();
      const color = series.color ? { Default: series.color } : colorRules.for(series) ?? (0, import_get_color_palette.getColorPalette)(
        Array.isArray(series.name) ? series.name[0] : series.name,
        startingIndex + i,
        colorPalette
      );
      const metadata = {
        id,
        color,
        gapPolicy: series.gapPolicy ?? gapPolicy,
        pointsDisplay: series.pointsDisplay ?? pointsDisplay
      };
      if (!valueRepPerVariant.get(variant) && series.valueRepresentation) {
        valueRepPerVariant.set(variant, series.valueRepresentation);
      }
      seriesMap.set(series, metadata);
    }
    if (!valueRepPerVariant.get(variant)) {
      valueRepPerVariant.set(variant, valueRepresentation);
    }
    startingIndex += variantSeries.length;
  }
  return {
    units: allUnits,
    seriesMap,
    valueRepPerVariant
  };
}
function cloneSeries(array) {
  return array.map((series) => ({
    ...series,
    datapoints: series.datapoints.map((datapoint) => ({ ...datapoint }))
  }));
}
const TimeseriesDataProvider = ({
  children,
  data: _data,
  valueRepresentation,
  onMessage
}) => {
  const clonedData = {
    line: cloneSeries(_data.line),
    bar: cloneSeries(_data.bar),
    area: cloneSeries(_data.area),
    band: cloneSeries(_data.band),
    shapePriority: _data.shapePriority
  };
  const data = (0, import_useDataByYAxis.useDataByYAxis)(clonedData);
  const mergedData = (0, import_react.useMemo)(() => (0, import_timeseries2.mergeTimeseriesFromVariants)(data), [data]);
  const { units, seriesMap, valueRepPerVariant } = useBuildInitialMetadata(
    data,
    valueRepresentation,
    mergedData
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_TimeseriesData_context.TimeseriesDataContext.Provider,
    {
      value: {
        data: mergedData,
        dataPerVariant: data,
        valueRepPerVariant,
        seriesMap,
        units
      },
      children
    }
  );
};
TimeseriesDataProvider["displayName"] = "TimeseriesDataProvider";
