// @dynatrace/strato-components/charts/timeseries/providers/TimeseriesFilteredDataProvider.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TimeseriesFilteredDataProvider_exports = {};
__export(TimeseriesFilteredDataProvider_exports, {
  TimeseriesFilteredDataProvider: () => TimeseriesFilteredDataProvider
});
module.exports = __toCommonJS(TimeseriesFilteredDataProvider_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_safe_get = require("../../core/utils/safe-get.js");
var import_TimeseriesFilteredData_context = require("../context/TimeseriesFilteredData.context.js");
var import_selectors = require("../hooks/store/selectors.js");
var import_useBuildFilteredMetadata = require("../hooks/useBuildFilteredMetadata.js");
var import_useTimeseriesData = require("../hooks/useTimeseriesData.js");
var import_timeseries2 = require("../utils/timeseries.js");
function binarySearchOrder(a, b) {
  return (0, import_timeseries2.getDatapointTime)(a.start, a.end) - (0, import_timeseries2.getDatapointTime)(b.start, b.end);
}
const TimeseriesFilteredDataProvider = ({ children }) => {
  const {
    dataPerVariant,
    seriesMap: initialSeriesMap,
    valueRepPerVariant,
    units
  } = (0, import_useTimeseriesData.useTimeseriesData)();
  const filteredSeriesIds = (0, import_selectors.useFilteredSeriesIdsState)();
  const hasId = (series) => filteredSeriesIds.has((0, import_safe_get.safeGet)(initialSeriesMap, series).id);
  const filteredTimeseriesPerVariant = filteredSeriesIds.size > 0 ? {
    ...dataPerVariant,
    line: dataPerVariant["line"].filter(hasId),
    bar: dataPerVariant["bar"].filter(hasId),
    area: dataPerVariant["area"].filter(hasId),
    band: dataPerVariant["band"].filter(hasId)
  } : dataPerVariant;
  const {
    seriesMap,
    datapointsMap,
    filteredUnits,
    filteredTimeseries,
    originalFilteredUnits
  } = (0, import_useBuildFilteredMetadata.useBuildFilteredMetadata)(
    filteredTimeseriesPerVariant,
    valueRepPerVariant,
    initialSeriesMap,
    units
  );
  const data = (0, import_timeseries2.mergeTimeseriesFromVariants)(filteredTimeseries);
  const allDatapoints = data.flatMap((series) => series.datapoints).sort(binarySearchOrder);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_TimeseriesFilteredData_context.TimeseriesFilteredDataContext.Provider,
    {
      value: {
        data,
        dataPerVariant: filteredTimeseries,
        seriesMap,
        datapointsMap,
        units: filteredUnits,
        originalUnits: originalFilteredUnits,
        allDatapoints
      },
      children
    }
  );
};
