// @dynatrace/strato-components/charts/core/components/sparkline-display/types/sparkline-container-height.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var sparkline_container_height_exports = {};
__export(sparkline_container_height_exports, {
  SparklineContainerHeight: () => SparklineContainerHeight,
  SparklineContainerHeightFontSize: () => SparklineContainerHeightFontSize
});
module.exports = __toCommonJS(sparkline_container_height_exports);
var SparklineContainerHeight = /* @__PURE__ */ ((SparklineContainerHeight2) => {
  SparklineContainerHeight2["HEIGHT24"] = "height24";
  SparklineContainerHeight2["HEIGHT32"] = "height32";
  return SparklineContainerHeight2;
})(SparklineContainerHeight || {});
var SparklineContainerHeightFontSize = /* @__PURE__ */ ((SparklineContainerHeightFontSize2) => {
  SparklineContainerHeightFontSize2["SMALLER_THAN_32"] = "smallerThan32";
  SparklineContainerHeightFontSize2["GREATER_OR_EQUAL_32"] = "greaterOrEqual32";
  return SparklineContainerHeightFontSize2;
})(SparklineContainerHeightFontSize || {});
