// @dynatrace/strato-components/charts/core/components/legend/range/components/RangeLegendSegmentsItem.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var RangeLegendSegmentsItem_exports = {};
__export(RangeLegendSegmentsItem_exports, {
  RangeLegendSegmentsItem: () => RangeLegendSegmentsItem
});
module.exports = __toCommonJS(RangeLegendSegmentsItem_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
var import_constants = require("../../../../constants.js");
var import_constants2 = require("../constants.js");
var import_RangeLegendOrientationContext = require("../context/RangeLegendOrientationContext.js");
var import_RangeLegendScaleContext = require("../context/RangeLegendScaleContext.js");
var import_use_legend_segment_keyboard_handler = require("../hooks/use-legend-segment-keyboard-handler.js");
var import_get_segment_item_position_and_size = require("../utils/get-segment-item-position-and-size.js");
var import_get_stroke_dasharray = require("../utils/get-stroke-dasharray.js");
const RangeLegendSegmentsItem = ({
  item,
  segmentSize,
  startPosition
}) => {
  const rectRef = (0, import_react.useRef)(null);
  const { orientation } = (0, import_RangeLegendOrientationContext.useOrientation)();
  const isVertical = orientation === import_constants.ORIENTATION.Vertical;
  const { scale } = (0, import_RangeLegendScaleContext.useScale)();
  const [, endScalePosition] = scale.range();
  const hasPreviousGap = startPosition > 0;
  const hasNextGap = Math.ceil(startPosition + segmentSize) < endScalePosition;
  const { itemHeight, itemWidth, itemX, itemY } = (0, import_get_segment_item_position_and_size.getSegmentItemPositionAndSize)(
    orientation,
    startPosition,
    segmentSize
  );
  const strokeDashArray = (0, import_get_stroke_dasharray.getStrokeDasharray)(
    { height: itemHeight, width: itemWidth },
    {
      bottom: isVertical && hasNextGap,
      left: !isVertical && hasPreviousGap,
      right: !isVertical && hasNextGap,
      top: isVertical && hasPreviousGap
    }
  );
  const { onBlurHandler, onFocusHandler, onKeyDownHandler } = (0, import_use_legend_segment_keyboard_handler.useLegendSegmentKeyboardHandler)(
    rectRef.current,
    item.from,
    item.to,
    startPosition
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "rect",
    {
      ref: rectRef,
      x: itemX,
      y: itemY,
      width: itemWidth,
      height: itemHeight,
      fill: item.color ?? import_constants2.DEFAULT_SEGMENT_COLOR,
      stroke: import_colors.default.Background.Base.Default,
      strokeWidth: import_constants2.DEFAULT_GAP_THICKNESS,
      strokeDasharray: strokeDashArray,
      role: "presentation",
      tabIndex: 0,
      onFocus: onFocusHandler,
      onBlur: onBlurHandler,
      onKeyDown: onKeyDownHandler
    }
  );
};
