// @dynatrace/strato-components/charts/core/components/legend/range/components/RangeLegendSegmentsOverlay.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var RangeLegendSegmentsOverlay_exports = {};
__export(RangeLegendSegmentsOverlay_exports, {
  RangeLegendSegmentsOverlay: () => RangeLegendSegmentsOverlay
});
module.exports = __toCommonJS(RangeLegendSegmentsOverlay_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_constants = require("../../../../constants.js");
var import_constants2 = require("../constants.js");
var import_RangeLegendOrientationContext = require("../context/RangeLegendOrientationContext.js");
var import_use_legend_segments_overlay_handlers = require("../hooks/use-legend-segments-overlay-handlers.js");
const RangeLegendSegmentsOverlay = (props) => {
  const { itemMarkers } = props;
  const overlayRef = (0, import_react.useRef)(null);
  const { onMouseLeaveHandler, onMouseMoveHandler } = (0, import_use_legend_segments_overlay_handlers.useLegendSegmentsOverlayHandlers)(itemMarkers);
  const { orientation } = (0, import_RangeLegendOrientationContext.useOrientation)();
  const isVertical = orientation === import_constants.ORIENTATION.Vertical;
  const overlaySizeHandler = (0, import_react.useCallback)(
    (legendBar, overlayEl) => {
      const { height: legendHeight, width: legendWidth } = legendBar.getBoundingClientRect();
      const overlayHeight = legendHeight + (isVertical ? 0 : import_constants2.OVERLAY_ADDITIONAL_SIZE);
      const overlayY = isVertical ? 0 : -import_constants2.OVERLAY_OFFSET;
      const overlayWidth = legendWidth + (isVertical ? import_constants2.OVERLAY_ADDITIONAL_SIZE : 0);
      const overlayX = isVertical ? -import_constants2.OVERLAY_OFFSET : 0;
      overlayEl.setAttribute("y", overlayY.toString());
      overlayEl.setAttribute("x", overlayX.toString());
      overlayEl.setAttribute("height", overlayHeight.toString());
      overlayEl.setAttribute("width", overlayWidth.toString());
    },
    [isVertical]
  );
  (0, import_react.useEffect)(() => {
    if (overlayRef.current) {
      const legendBar = overlayRef.current.closest("svg")?.previousElementSibling;
      if (legendBar) {
        overlaySizeHandler(legendBar, overlayRef.current);
        const observer = new MutationObserver(() => {
          overlaySizeHandler(legendBar, overlayRef.current);
        });
        observer.observe(legendBar, {
          attributes: true,
          childList: true,
          subtree: true,
          characterData: true
        });
        return () => observer.disconnect();
      }
    }
  }, [overlaySizeHandler]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "rect",
    {
      ref: overlayRef,
      role: "presentation",
      id: "segments-overlay",
      fill: "transparent",
      onMouseMove: onMouseMoveHandler,
      onMouseLeave: onMouseLeaveHandler
    }
  );
};
