// @dynatrace/strato-components/esm/charts/core/components/legend/categorical/components/ChartLegendSeriesActions/ChartLegendSeriesActions.js
import { jsx, jsxs } from "react/jsx-runtime";
import { DotMenuIcon } from "@dynatrace/strato-icons";
import { seriesButton } from "./ChartLegendSeriesActions.sty.js";
import { Button } from "../../../../../../../buttons/button/Button.js";
import { Menu } from "../../../../../../../navigation/menu/Menu.js";
import { useCustomSeriesActions } from "../../../../../hooks/useCustomSeriesActions.js";
import { renderMenuActionChildren } from "../../../../../utils/render-menu-action-children.js";
const ChartLegendSeriesActions = (props) => {
  const {
    disabled,
    onOpenChange,
    isOpen,
    visible = false,
    onFocus,
    onBlur,
    onMouseEnter,
    onMouseLeave,
    defaultActions,
    actions
  } = props;
  const handleFocus = (event) => {
    event.stopPropagation();
    onFocus();
  };
  const customActions = useCustomSeriesActions(actions, isOpen);
  return /* @__PURE__ */ jsxs(Menu, { onOpenChange, children: [
    /* @__PURE__ */ jsx(Menu.Trigger, { children: /* @__PURE__ */ jsx(
      Button,
      {
        "data-testid": "legend-more-actions-button",
        variant: "default",
        disabled,
        onFocus: handleFocus,
        onBlur,
        onMouseEnter,
        onMouseLeave,
        className: seriesButton({ visible }),
        children: /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(DotMenuIcon, { size: "small" }) })
      }
    ) }),
    !disabled && isOpen && /* @__PURE__ */ jsxs(Menu.Content, { "data-testid": "menu-content", children: [
      defaultActions ? renderMenuActionChildren(defaultActions) : null,
      customActions ? /* @__PURE__ */ jsx("div", { "data-dt-children-name": "LegendSeriesItems", children: renderMenuActionChildren(customActions.props.children) }) : null
    ] })
  ] });
};
ChartLegendSeriesActions["displayName"] = "LegendSeriesActions";
export {
  ChartLegendSeriesActions
};
