// @dynatrace/strato-components/esm/charts/gauge/constants.js
import Colors from "@dynatrace/strato-design-tokens/colors";
const DEFAULT_MIN_RANGE_VALUE = 0;
const DEFAULT_MAX_RANGE_VALUE = 100;
const DEFAULT_CHART_WIDTH = "100%";
const DEFAULT_CHART_HEIGHT = 300;
const MIN_SIZE_IN_PX = 100;
const MIN_MAX_PADDING = 20;
const MIN_MAX_LABEL_RESERVED_HEIGHT_IN_PX = 16;
const ARC_OUTER_RADIUS_PERCENTAGE = 0.9;
const ARC_INNER_RADIUS_PERCENTAGE = 0.8;
const GAUGE_VERTICAL_SMALL_OFFSET = 0.05;
const GAUGE_VERTICAL_DEFAULT_OFFSET = 0.1;
const ARC_BACKGROUND_COLOR = Colors.Theme.Neutral["40"];
const ARC_DEFAULT_VALUE_COLOR = Colors.Charts.Categorical.Color11.Default;
const THRESHOLD_COLOR_DEFAULT = Colors.Charts.Categorical.Color11.Default;
const ARC_START_ANGLE = -0.625 * Math.PI;
const ARC_END_ANGLE = 0.625 * Math.PI;
const THRESHOLD_LINE_WIDTH = 2;
const THRESHOLD_LINE_DEFAULT_COLOR = Colors.Icon.Neutral.Default;
const GAUGE_DEFAULT_CONFIG = {
  color: ARC_DEFAULT_VALUE_COLOR,
  min: DEFAULT_MIN_RANGE_VALUE,
  max: DEFAULT_MAX_RANGE_VALUE,
  formatter: void 0,
  loading: false,
  tooltip: {}
};
const DISPLAY_NAME = "GaugeChart";
const GAUGE_CHART_ARIA_LABEL = "Gauge Chart";
const FONT_HEIGHT_RATIO = 0.7;
const DEFAULT_MAX_VALUE_FONT_SIZE = 240;
const zIndexGaugeAreas = {
  toolbar: 2,
  tooltip: 1,
  overlay: 0
};
export {
  ARC_BACKGROUND_COLOR,
  ARC_DEFAULT_VALUE_COLOR,
  ARC_END_ANGLE,
  ARC_INNER_RADIUS_PERCENTAGE,
  ARC_OUTER_RADIUS_PERCENTAGE,
  ARC_START_ANGLE,
  DEFAULT_CHART_HEIGHT,
  DEFAULT_CHART_WIDTH,
  DEFAULT_MAX_RANGE_VALUE,
  DEFAULT_MAX_VALUE_FONT_SIZE,
  DEFAULT_MIN_RANGE_VALUE,
  DISPLAY_NAME,
  FONT_HEIGHT_RATIO,
  GAUGE_CHART_ARIA_LABEL,
  GAUGE_DEFAULT_CONFIG,
  GAUGE_VERTICAL_DEFAULT_OFFSET,
  GAUGE_VERTICAL_SMALL_OFFSET,
  MIN_MAX_LABEL_RESERVED_HEIGHT_IN_PX,
  MIN_MAX_PADDING,
  MIN_SIZE_IN_PX,
  THRESHOLD_COLOR_DEFAULT,
  THRESHOLD_LINE_DEFAULT_COLOR,
  THRESHOLD_LINE_WIDTH,
  zIndexGaugeAreas
};
