// @dynatrace/strato-components/esm/charts/gauge/GaugeChart.js
import { jsx } from "react/jsx-runtime";
import {
  forwardRef,
  useImperativeHandle,
  useRef
} from "react";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useMergeRefs } from "../../core/hooks/useMergeRefs.js";
import { useObserveContainerSize } from "../core/hooks/useObserveContainerSize.js";
import { ChartType } from "../core/types/chart-types.js";
import { coerceSizeValue } from "../core/utils/coerceSizeValue.js";
import { GaugeChartLayout } from "./components/GaugeChartLayout/GaugeChartLayout.js";
import {
  DEFAULT_CHART_HEIGHT,
  DEFAULT_CHART_WIDTH,
  DISPLAY_NAME,
  GAUGE_CHART_ARIA_LABEL
} from "./constants.js";
import { useGaugeChartConfig } from "./hooks/useGaugeChartConfig.js";
import { GaugeChartProviders } from "./providers/GaugeChartProviders.js";
import { Label } from "./slots/Label.js";
import { ThresholdIndicator } from "./slots/ThresholdIndicator.js";
import { GaugeChartToolbarSlot } from "./slots/Toolbar.js";
import { Tooltip } from "./slots/Tooltip.js";
import { Intent } from "../../core/slots/Intent/Intent.js";
import { ColorRule } from "../core/slots/ColorRule.js";
const COMPONENT_NAME = "GaugeChart";
const _GaugeChart = forwardRef((props, forwardedRef) => {
  const {
    value,
    width,
    height,
    style,
    className,
    prefixIcon,
    seriesActions,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "data-testid": dataTestId = "gauge-chart-root",
    valueAccessor
  } = props;
  const [behavioralTrackingProps] = useBehavioralTrackingProps(props);
  const config = useGaugeChartConfig(props);
  const {
    min,
    max,
    color,
    formatter,
    thresholdsIndicators,
    colorRules,
    toolbar,
    intents,
    tooltip,
    valueAccessor: configValueAccessor,
    loading
  } = config;
  const effectiveValueAccessor = valueAccessor ?? configValueAccessor;
  const chartWidth = coerceSizeValue(width) ?? DEFAULT_CHART_WIDTH;
  const chartHeight = coerceSizeValue(height) ?? DEFAULT_CHART_HEIGHT;
  const [{ height: containerHeight, width: containerWidth }, containerRef] = useObserveContainerSize();
  const gaugeChartRef = useRef(null);
  useImperativeHandle(forwardedRef, () => {
    const consumerConfig = {
      min,
      max,
      color,
      formatter,
      thresholdsIndicators,
      colorRules,
      toolbar,
      tooltip,
      valueAccessor: effectiveValueAccessor,
      loading
    };
    return {
      element: gaugeChartRef.current,
      getConfig: () => JSON.stringify({ type: ChartType.GaugeChart, ...consumerConfig })
    };
  });
  const ref = useMergeRefs([
    forwardedRef,
    gaugeChartRef
  ]);
  return /* @__PURE__ */ jsx(
    "div",
    {
      ...behavioralTrackingProps,
      ref: containerRef,
      style: {
        width: chartWidth,
        height: chartHeight,
        overflow: "hidden",
        ...style
      },
      className,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "data-testid": dataTestId,
      role: "img",
      "aria-label": GAUGE_CHART_ARIA_LABEL,
      "data-dt-component": COMPONENT_NAME,
      children: /* @__PURE__ */ jsx(
        GaugeChartProviders,
        {
          config,
          width: containerWidth,
          height: containerHeight,
          value,
          valueAccessor: effectiveValueAccessor,
          explicitColor: color,
          seriesActions,
          loading: loading ?? false,
          children: /* @__PURE__ */ jsx(
            GaugeChartLayout,
            {
              ref,
              value,
              valueAccessor: effectiveValueAccessor,
              prefixIcon,
              intents,
              toolbar
            }
          )
        }
      )
    }
  );
});
_GaugeChart.displayName = DISPLAY_NAME;
const GaugeChart = Object.assign(_GaugeChart, {
  ColorRule,
  Intent,
  Label,
  ThresholdIndicator,
  Toolbar: GaugeChartToolbarSlot,
  Tooltip
});
export {
  GaugeChart
};
