// @dynatrace/strato-components/content/feature-highlight/FeatureHighlight.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var FeatureHighlight_exports = {};
__export(FeatureHighlight_exports, {
  FeatureHighlight: () => FeatureHighlight
});
module.exports = __toCommonJS(FeatureHighlight_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_core = require("@popperjs/core");
var import_react = require("react");
var import_react_dom = require("react-dom");
var import_react_intl = require("react-intl");
var import_react_popper = require("react-popper");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_Actions = require("./Actions.js");
var import_FeatureHighlight_css = require("./FeatureHighlight.sty.js");
var import_Button = require("../../buttons/button/Button.js");
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_useOverlayStacking = require("../../core/hooks/useOverlayStacking.js");
var import_useSafeForwardProps = require("../../core/hooks/useSafeForwardProps.js");
var import_is_slot = require("../../core/utils/is-slot.js");
var import_logging = require("../../core/utils/logging.js");
var import_Flex = require("../../layouts/flex/Flex.js");
var import_useKeyboardFocusHandling = require("../../overlays/hooks/useKeyboardFocusHandling.js");
var import_placement = require("../../overlays/types/placement.js");
var import_popper_modifiers = require("../../overlays/utils/popper-modifiers.js");
var import_Content = require("../shared-components/Content.js");
var import_Title = require("../shared-components/Title.js");
var import_Visual = require("../shared-components/Visual.js");
const COMPONENT_NAME = "FeatureHighlight";
const getPreferredPlacement = (placement) => {
  const preferredPlacement = placement.split("-")[0];
  if (import_placement.basePlacements.includes(preferredPlacement)) {
    return preferredPlacement;
  } else {
    return import_placement.basePlacements[1];
  }
};
const FeatureHighlightRoot = (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const {
      open,
      anchor,
      placement = "right-bottom",
      onClose,
      children,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      className: consumerClassName,
      style: consumerStyle,
      ...remainingProps
    } = props;
    const preferredPlacement = getPreferredPlacement(placement);
    const [overlayElement, setOverlayElement] = (0, import_react.useState)();
    const [arrowElement, setArrowElement] = (0, import_react.useState)();
    const closeButtonRef = (0, import_react.useRef)(null);
    const featureHighlightContainer = (0, import_react.useRef)(null);
    const rootRef = (0, import_useMergeRefs.useMergeRefs)([
      setOverlayElement,
      forwardedRef
    ]);
    const [handleFocusOnKeydown, closedOnKeydown] = (0, import_useKeyboardFocusHandling.useKeyboardFocusHandling)(closeButtonRef);
    const overlayStacking = (0, import_useOverlayStacking.useOverlayStacking)(void 0, "feature-highlight");
    const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
    const ariaLabelingProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(restProps);
    const forwardProps = (0, import_useSafeForwardProps.useSafeForwardProps)(restProps, ariaLabelingProps);
    const intl = (0, import_react_intl.useIntl)();
    const slots = {
      title: null,
      visual: null,
      content: null,
      actions: null
    };
    (0, import_react.useLayoutEffect)(() => {
      const toastContainer = document.querySelector("[data-toast-container]");
      const microguideContainer = document.querySelector(
        "[data-microguide-container]"
      );
      featureHighlightContainer.current = document.createElement("div");
      featureHighlightContainer.current.setAttribute("aria-live", "polite");
      featureHighlightContainer.current.setAttribute(
        "data-feature-highlight-container",
        "true"
      );
      if (microguideContainer && microguideContainer.parentNode) {
        microguideContainer.parentNode.insertBefore(
          featureHighlightContainer.current,
          microguideContainer
        );
      } else if (toastContainer && toastContainer.parentNode) {
        toastContainer.parentNode.insertBefore(
          featureHighlightContainer.current,
          toastContainer
        );
      } else {
        document.body.appendChild(featureHighlightContainer.current);
      }
      return () => {
        featureHighlightContainer.current?.remove();
      };
    }, []);
    import_react.Children.forEach(children, (child) => {
      if ((0, import_is_slot.isSlot)(child, import_Title.Title)) {
        slots.title = child;
      } else if ((0, import_is_slot.isSlot)(child, import_Visual.Visual)) {
        slots.visual = child;
      } else if ((0, import_is_slot.isSlot)(child, import_Content.Content)) {
        slots.content = child;
      } else if ((0, import_is_slot.isSlot)(child, import_Actions.Actions)) {
        slots.actions = child;
      } else if ((0, import_react.isValidElement)(child)) {
        (0, import_logging.devLog)(
          "Only Title, Visual, Content and Actions are valid children of FeatureHighlight."
        );
      } else if (child !== null) {
        (0, import_logging.devLog)("The used child is not supported by FeatureHighlight.");
      }
    });
    if (slots.title && ariaLabelingProps["aria-labelledby"] === void 0) {
      ariaLabelingProps["aria-labelledby"] = "strato-feature-highlight-title";
    }
    if (slots.content && ariaLabelingProps["aria-describedby"] === void 0) {
      ariaLabelingProps["aria-describedby"] = "strato-feature-highlight-content";
    }
    const closeOnEscape = (e) => {
      if (e.key === "Escape") {
        e.preventDefault();
        closedOnKeydown.current = true;
        onClose();
      }
    };
    const offsetModifier = (0, import_react.useMemo)(
      () => ({
        name: "offset",
        options: {
          offset: (({ placement: popperPlacement, popper }) => {
            let skidding = 0;
            const spacingToPopperEdge = 24;
            if (popperPlacement === preferredPlacement) {
              switch (placement) {
                case "right-bottom":
                case "left-bottom":
                  skidding = popper.height / 2 - spacingToPopperEdge;
                  break;
                case "right-top":
                case "left-top":
                  skidding = -(popper.height / 2 - spacingToPopperEdge);
                  break;
                case "top-left":
                case "bottom-left":
                  skidding = -(popper.width / 2 - spacingToPopperEdge);
                  break;
                case "top-right":
                case "bottom-right":
                  skidding = popper.width / 2 - spacingToPopperEdge;
                  break;
                default:
                  skidding = 0;
              }
            }
            return [skidding, 16];
          })
        }
      }),
      [placement, preferredPlacement]
    );
    const { styles, attributes, update } = (0, import_react_popper.usePopper)(anchor, overlayElement, {
      placement: preferredPlacement,
      modifiers: [
        offsetModifier,
        {
          name: "flip",
          options: {
            fallbackPlacements: [...import_placement.basePlacements]
          }
        },
        {
          name: "arrow",
          options: {
            element: arrowElement,
            padding: 16
          }
        },
        import_popper_modifiers.observeOverlaySize,
        import_popper_modifiers.observeAnimationDone,
        import_core.eventListeners
      ]
    });
    (0, import_react.useEffect)(() => {
      if (!anchor || !open || !update) {
        return;
      }
      let prevRect = anchor.getBoundingClientRect();
      let positionPollRequestId = null;
      let pendingAnimationFrame = null;
      let isUpdateScheduled = false;
      const scheduleUpdate = () => {
        if (isUpdateScheduled) {
          return;
        }
        isUpdateScheduled = true;
        if (pendingAnimationFrame !== null) {
          cancelAnimationFrame(pendingAnimationFrame);
        }
        pendingAnimationFrame = requestAnimationFrame(async () => {
          await update();
          isUpdateScheduled = false;
          pendingAnimationFrame = null;
        });
      };
      const pollForPositionChange = () => {
        const anchorRect = anchor.getBoundingClientRect();
        const hasPositionChanged = Math.abs(anchorRect.x - prevRect.x) > 0.5 || Math.abs(anchorRect.y - prevRect.y) > 0.5 || Math.abs(anchorRect.width - prevRect.width) > 0.5 || Math.abs(anchorRect.height - prevRect.height) > 0.5;
        if (hasPositionChanged) {
          prevRect = anchorRect;
          scheduleUpdate();
        }
        positionPollRequestId = requestAnimationFrame(pollForPositionChange);
      };
      positionPollRequestId = requestAnimationFrame(pollForPositionChange);
      return () => {
        if (positionPollRequestId !== null) {
          cancelAnimationFrame(positionPollRequestId);
        }
        if (pendingAnimationFrame !== null) {
          cancelAnimationFrame(pendingAnimationFrame);
        }
      };
    }, [anchor, open, update]);
    (0, import_react.useEffect)(() => {
      if (open === false) {
        handleFocusOnKeydown();
      }
    }, [handleFocusOnKeydown, open]);
    return open && anchor ? (0, import_react_dom.createPortal)(
      /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
        import_Flex.Flex,
        {
          ref: rootRef,
          onKeyDown: closeOnEscape,
          role: "dialog",
          "aria-modal": "false",
          className: consumerClassName,
          style: {
            ...styles.popper,
            zIndex: overlayStacking.style.zIndex,
            ...consumerStyle
          },
          "data-testid": dataTestId,
          "data-dtrum-mask": dataDtrumMask,
          "data-dtrum-allow": dataDtrumAllow,
          ...attributes.popper,
          ...ariaLabelingProps,
          ...behavioralTrackingProps,
          ...forwardProps,
          "data-dt-component": COMPONENT_NAME,
          children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
              import_Flex.Flex,
              {
                className: import_FeatureHighlight_css.FeatureHighlightScrollContainerCSS,
                flexDirection: "column",
                gap: 0,
                paddingX: 16,
                paddingTop: 8,
                paddingBottom: 8,
                children: [
                  /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
                    import_Flex.Flex,
                    {
                      flexDirection: "row",
                      justifyContent: "space-between",
                      alignItems: "flex-start",
                      gap: 8,
                      paddingBottom: slots.content !== null || slots.actions !== null || slots.visual !== null ? 8 : 0,
                      children: [
                        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                          import_Flex.Flex,
                          {
                            flexItem: true,
                            flexGrow: 1,
                            paddingY: 6,
                            id: "strato-feature-highlight-title",
                            children: slots.title
                          }
                        ),
                        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                          import_Button.Button,
                          {
                            className: import_FeatureHighlight_css.FeatureHighlightCloseButtonCSS,
                            onClick: () => onClose(),
                            onKeyDown: () => closedOnKeydown.current = true,
                            ref: closeButtonRef,
                            "aria-label": intl.formatMessage({
                              defaultMessage: "Close feature highlight",
                              id: "XQHEJyt2t/js0TGr",
                              description: "Close the feature highlight."
                            }),
                            "data-dt-name": "CloseX",
                            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.XmarkIcon, {}) })
                          }
                        )
                      ]
                    }
                  ),
                  slots.visual && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                    import_Flex.Flex,
                    {
                      paddingBottom: slots.content !== null || slots.actions !== null ? 16 : 8,
                      children: slots.visual
                    }
                  ),
                  slots.content && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                    import_Flex.Flex,
                    {
                      paddingBottom: 8,
                      style: { overflow: "hidden" },
                      id: "strato-feature-highlight-content",
                      children: slots.content
                    }
                  ),
                  slots.actions
                ]
              }
            ),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              "div",
              {
                "data-popper-arrow": true,
                ref: setArrowElement,
                className: import_FeatureHighlight_css.FeatureHighlightPointerArrowCSS,
                style: styles.arrow,
                ...attributes.arrow
              }
            )
          ]
        }
      ),
      featureHighlightContainer.current || document.body
    ) : null;
  }
);
FeatureHighlightRoot.displayName = COMPONENT_NAME;
const FeatureHighlight = Object.assign(FeatureHighlightRoot, {
  Title: import_Title.Title,
  Visual: import_Visual.Visual,
  Content: import_Content.Content,
  Actions: import_Actions.Actions
});
