// @dynatrace/strato-components/tables/DataTable/features/ColumnTypes/ColumnTypes.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ColumnTypes_exports = {};
__export(ColumnTypes_exports, {
  applyColumnTypes: () => applyColumnTypes,
  getAlignmentBasedOnType: () => getAlignmentBasedOnType,
  isDatatableGanttColumnConfig: () => isDatatableGanttColumnConfig,
  isDatatableLogContentColumnConfig: () => isDatatableLogContentColumnConfig
});
module.exports = __toCommonJS(ColumnTypes_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
var import_units = require("@dynatrace-sdk/units");
var import_ColumnTypes_css = require("./ColumnTypes.sty.js");
var import_Sparkline = require("../../../../charts/sparkline/Sparkline.js");
var import_Markdown = require("../../../../content/markdown/Markdown.js");
var import_default_formatters = require("../../../../core/utils/default-formatters.js");
var import_logging = require("../../../../core/utils/logging.js");
var import_string_utils = require("../../../../core/utils/string-utils.js");
var import_log_content_highlighting = require("../../../utils/log-content-highlighting.js");
var import_replace_urls_with_links = require("../../../utils/replace-urls-with-links.js");
var import_constants = require("../../constants.js");
var import_DataTableGanttChartCell = require("../GanttChart/DataTableGanttChartCell.js");
var import_DataTableGanttChartHeader = require("../GanttChart/DataTableGanttChartHeader.js");
var import_MeterBarChart = require("../MeterBarChart/MeterBarChart.js");
function applyColumnTypes(columnDef, composedDef) {
  if (columnDef.cell !== void 0) {
    return {};
  }
  const columnType = columnDef.columnType;
  if (!columnType) {
    return {};
  }
  if (columnType === "date") {
    (0, import_logging.devLog)(
      "DataTable: columnType 'date' is deprecated and will be removed in a future release. Use 'datetime' instead."
    );
  }
  let enableUserActions = true;
  if (columnType === "sparkline") {
    columnDef.disableSorting = true;
  }
  if (columnType === "meterbar") {
    columnDef.disableSorting ??= true;
    enableUserActions = { type: "chart" };
  }
  if (columnType === "markdown") {
    columnDef.disableSorting = true;
  }
  if (columnType === "gantt") {
    columnDef.disableSorting = true;
    enableUserActions = false;
    columnDef.header = () => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_DataTableGanttChartHeader.DataTableGanttChartHeader, { columnDef });
    composedDef.meta = {
      ...composedDef.meta,
      hasCustomCellRenderer: true,
      hasCustomHeaderRenderer: true
    };
  }
  if (columnType === "log-content") {
    composedDef.meta = {
      ...composedDef.meta,
      fontStyle: "code"
    };
  }
  return {
    alignment: columnDef.alignment || getAlignmentBasedOnType(columnType),
    cell: getCellTypeRenderer(columnType),
    enableSorting: !columnDef.disableSorting,
    header: columnDef.header,
    enableUserActions,
    meta: composedDef.meta
  };
}
function getAlignmentBasedOnType(columnType) {
  switch (columnType) {
    case "number":
    case "long":
    case "currency":
    case "bit":
    case "date":
    // DEPRECATION: [APPDEV-16044] BREAKING-CHANGE: Remove 'date' columnType.
    case "datetime":
      return "right";
    default:
      return import_constants.DATA_TABLE_DEFAULT_HORIZONTAL_ALIGNMENT;
  }
}
function isDatatableGanttColumnConfig(config) {
  return config !== null && config !== void 0 && typeof config === "object" && "nameAccessor" in config;
}
function isDatatableSparklineColumnConfig(config) {
  return config !== null && config !== void 0 && typeof config === "object" && ("color" in config || "variant" in config || "gapPolicy" in config || "curve" in config || "showContextValues" in config || "xAxis" in config || "yAxis" in config);
}
function isDatatableMarkdownColumnConfig(config) {
  return config !== null && config !== void 0 && typeof config === "object" && "customComponentMappings" in config;
}
function isDatatableLogContentColumnConfig(config) {
  return config !== null && config !== void 0 && typeof config === "object" && "truncationLimit" in config;
}
function isDatatableTextColumnConfig(config) {
  return config !== null && config !== void 0 && typeof config === "object" && "detectLinks" in config;
}
const SparklineMemo = (0, import_react.memo)(import_Sparkline.Sparkline);
const MeterbarMemo = (0, import_react.memo)(import_MeterBarChart.MeterBarChartRenderer);
const MarkdownMemo = (0, import_react.memo)(import_Markdown.Markdown);
const DatetimeMemo = (0, import_react.memo)((props) => {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, { children: (0, import_default_formatters.formatDatetimeDefault)(
    new Date((0, import_string_utils.convertToNumberIfPureNumber)(props.value))
  ) });
});
const CELL_TYPE_RENDERERS = {
  number: (props) => {
    const cellValue = props.getValue();
    if (typeof cellValue === "string" || typeof cellValue === "number") {
      const valueAsNumber = Number.isNaN(+cellValue) || cellValue === null ? "" : (0, import_string_utils.convertToNumberIfPureNumber)(cellValue);
      return typeof valueAsNumber === "number" ? (0, import_units.adjustFractionDigits)(valueAsNumber) : "";
    }
  },
  long: (props) => {
    const cellValue = props.getValue();
    if (cellValue === void 0 || cellValue === null || Number.isNaN(cellValue)) {
      return "";
    } else if (typeof cellValue === "number" || typeof cellValue === "bigint" || typeof cellValue === "string") {
      return (0, import_units.formatLong)(cellValue);
    }
    return `${cellValue}`;
  },
  currency: (props) => {
    const cellValue = props.getValue();
    if (typeof cellValue === "string" || typeof cellValue === "number") {
      const valueAsNumber = Number.isNaN(+cellValue) || cellValue === null ? "" : (0, import_string_utils.convertToNumberIfPureNumber)(cellValue);
      return typeof valueAsNumber === "number" ? (
        // no currency (e.g. 'EUR', 'USD') provided → default format without currency symbol
        (0, import_default_formatters.formatCurrencyDefault)(valueAsNumber)
      ) : "";
    }
  },
  bit: (props) => {
    const cellValue = props.getValue();
    if (typeof cellValue === "string" || typeof cellValue === "number") {
      const valueAsNumber = Number.isNaN(+cellValue) ? "" : (0, import_string_utils.convertToNumberIfPureNumber)(cellValue);
      return typeof valueAsNumber === "number" ? (0, import_units.format)(valueAsNumber, { input: import_units.units.data.bit }) : "";
    }
  },
  // DEPRECATION: [APPDEV-16044] BREAKING-CHANGE: Remove 'date' columnType.
  date: (props) => {
    const cellValue = props.getValue();
    if (cellValue === null) {
      return null;
    }
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(DatetimeMemo, { value: cellValue });
  },
  datetime: (props) => {
    const cellValue = props.getValue();
    if (cellValue === null) {
      return null;
    }
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(DatetimeMemo, { value: cellValue });
  },
  text: (props) => {
    const cellValue = props.getValue();
    let detectLinks = true;
    if (isDatatableTextColumnConfig(props.column.columnDef.meta?.config)) {
      detectLinks = !!props.column.columnDef.meta?.config.detectLinks;
    }
    if (typeof cellValue !== "string" || !detectLinks) {
      return cellValue === void 0 || cellValue === null || Number.isNaN(cellValue) ? "" : `${cellValue}`;
    }
    return (0, import_replace_urls_with_links.replaceUrlsWithLinks)(cellValue);
  },
  sparkline: (props) => {
    let config = {};
    if (props.column.columnDef.meta && isDatatableSparklineColumnConfig(props.column.columnDef.meta.config)) {
      config = props.column.columnDef.meta.config;
    }
    const rowDensity = props.table.getVisualVariantOptions().rowDensity;
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      SparklineMemo,
      {
        data: props.getValue(),
        color: config.color ?? import_colors.default.Charts.Categorical.Color01.Default,
        variant: config.variant ?? "line",
        gapPolicy: config.gapPolicy ?? "connect",
        curve: config.curve ?? "linear",
        showContextValues: !!config?.showContextValues,
        labelsAlignment: "right",
        labelsFixedWidth: rowDensity === "condensed" ? 36 : 42,
        children: [
          config.xAxis && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Sparkline.Sparkline.XAxis, { ...config.xAxis }),
          config.yAxis && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Sparkline.Sparkline.YAxis, { ...config.yAxis })
        ]
      }
    );
  },
  meterbar: (props) => {
    const value = props.getValue();
    if (typeof value === "number" || (0, import_MeterBarChart.isMultiMeterBarData)(value)) {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(MeterbarMemo, { value, column: props.column });
    }
    return "";
  },
  gantt: (props) => {
    let config = { nameAccessor: "name" };
    if (props.column.columnDef.meta && isDatatableGanttColumnConfig(props.column.columnDef.meta.config)) {
      config = props.column.columnDef.meta.config;
    }
    const value = props.getValue();
    const parentData = props.row.getParentRow()?.getValue(props.column.id);
    const isGanttValue = (0, import_DataTableGanttChartCell.isGanttSegmentData)(value) || Array.isArray(value) && value.every(import_DataTableGanttChartCell.isGanttSegmentData);
    const isGanttParentData = parentData === void 0 || (0, import_DataTableGanttChartCell.isGanttSegmentData)(parentData) || Array.isArray(parentData) && parentData.every(import_DataTableGanttChartCell.isGanttSegmentData);
    const pageRows = props.table.getPaginationRowModel().rows;
    const isLastPageRow = props.row.id === pageRows.at(-1)?.id;
    if (isGanttValue && isGanttParentData) {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_DataTableGanttChartCell.DataTableGanttChartCell,
        {
          data: value,
          parentData,
          columnId: props.column.id,
          config,
          isLastPageRow
        }
      );
    }
    return "";
  },
  markdown: (props) => {
    const config = props.column.columnDef.meta?.config;
    const customComponentMappings = isDatatableMarkdownColumnConfig(config) ? config?.customComponentMappings : void 0;
    const value = props.getValue();
    if (value !== null && value !== void 0 && typeof value === "string") {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(MarkdownMemo, { customComponentMappings, children: value });
    }
    return "";
  },
  "log-content": (props) => {
    const cellValue = props.getValue();
    if (typeof cellValue === "string") {
      const isColumnLineWrap = props.column.getIsLineWrapped();
      const config = props.column.columnDef.meta?.config;
      const truncationLimit = isDatatableLogContentColumnConfig(config) ? config?.truncationLimit : void 0;
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: (0, import_ColumnTypes_css.logContentWrapper)({ linewrap: isColumnLineWrap }), children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_log_content_highlighting.LogContentHighlight,
        {
          logContent: cellValue,
          lineWrap: isColumnLineWrap,
          truncationLimit
        }
      ) });
    }
    return "";
  }
};
function getCellTypeRenderer(columnType) {
  return CELL_TYPE_RENDERERS[columnType] ?? CELL_TYPE_RENDERERS.text;
}
