// @dynatrace/strato-components/esm/charts/core/components/annotations/components/canvas/displays/AnnotationsMarkersDisplay.js
import { jsx, jsxs } from "react/jsx-runtime";
import { Fragment, useContext } from "react";
import { useCurrentTheme } from "../../../../../../../core/hooks/useCurrentTheme.js";
import { getTokenComputedStyle } from "../../../../../utils/getTokenComputedStyle.js";
import { RectClipPathCanvas } from "../../../../canvas/components/RectClipPathCanvas.js";
import {
  MARKER_STROKE_COLOR,
  MARKER_STROKE_WIDTH,
  MINIMUM_MARKER_WIDTH,
  PILL_MARKER_HEIGHT,
  TRACK_HEIGHT
} from "../../../constants.js";
import { AnnotationsConfigContext } from "../../../contexts/AnnotationsConfig.context.js";
import { cropMarker } from "../../../utils/crop-marker.js";
import { getMarkerOffsetByTrack } from "../../../utils/get-markers-offset-by-track.js";
import { useTrackDisplay } from "../hooks/use-track-display.js";
import { CanvasContextRestore } from "../renderers/CanvasContextRestore.js";
import { RangeMarkerRenderer } from "../renderers/RangeMarkerRenderer.js";
import { VALUE_MARKER_WIDTH } from "../renderers/value-marker-renderer/constants.js";
import {
  ValueMarkerRenderer
} from "../renderers/value-marker-renderer/ValueMarkerRenderer.js";
import { calculateMarkerWidthAndContent } from "../utils/calculate-marker-width-and-content.js";
import { getMarkerBrightnessPercentageBasedOnState } from "../utils/get-marker-brightness-percentage-based-on-state.js";
const MARKER_DEFAULT_OPACITY = 0.85;
const MARKER_HOVER_OPACITY = 0.2;
const AnnotationsMarkersDisplay = ({
  markers
}) => {
  const theme = useCurrentTheme();
  const { trackWidth } = useContext(AnnotationsConfigContext);
  const { offsetX, trackIndex } = useTrackDisplay();
  return markers.map((props, markerKey) => {
    const {
      color,
      content,
      interactionState = "default",
      width: widthProp,
      x,
      textOverflow,
      truncateMode
    } = props;
    const isRange = widthProp > 0;
    const outOfBounds = isRange ? x + widthProp <= 0 : x < 0 || x > trackWidth;
    if (outOfBounds) {
      return null;
    }
    const markerMinWidth = isRange ? MINIMUM_MARKER_WIDTH : VALUE_MARKER_WIDTH;
    const width = Math.max(widthProp, markerMinWidth);
    const markerOffsetByTrack = getMarkerOffsetByTrack(trackIndex);
    const brightness = getMarkerBrightnessPercentageBasedOnState(
      theme,
      interactionState
    );
    const opacity = interactionState === "inactive" ? MARKER_HOVER_OPACITY : MARKER_DEFAULT_OPACITY;
    const { croppedX, croppedWidth } = cropMarker(x, width, trackWidth);
    const {
      sanitizedContent,
      calculatedMarkerWidth,
      valueMarkerOrientation,
      calculatedX
    } = calculateMarkerWidthAndContent({
      trackWidth,
      trackLeftBoundary: offsetX,
      content,
      width: croppedWidth,
      x: croppedX,
      isRange,
      textOverflow,
      truncateMode
    });
    const markerProps = {
      x: calculatedX,
      y: markerOffsetByTrack,
      width: calculatedMarkerWidth,
      height: PILL_MARKER_HEIGHT,
      fillColor: getTokenComputedStyle(color),
      strokeColor: getTokenComputedStyle(MARKER_STROKE_COLOR),
      opacity,
      brightness,
      children: sanitizedContent
    };
    return isRange ? /* @__PURE__ */ jsxs(Fragment, { children: [
      /* @__PURE__ */ jsx(
        RectClipPathCanvas,
        {
          position: {
            x: offsetX - MARKER_STROKE_WIDTH * devicePixelRatio,
            y: markerOffsetByTrack - MARKER_STROKE_WIDTH * devicePixelRatio
          },
          height: TRACK_HEIGHT,
          width: trackWidth + 2 * MARKER_STROKE_WIDTH * devicePixelRatio
        }
      ),
      /* @__PURE__ */ jsx(RangeMarkerRenderer, { ...markerProps }),
      /* @__PURE__ */ jsx(CanvasContextRestore, {})
    ] }, `${markerKey}_${x}`) : /* @__PURE__ */ jsx(Fragment, { children: /* @__PURE__ */ jsx(
      ValueMarkerRenderer,
      {
        ...markerProps,
        orientation: valueMarkerOrientation
      }
    ) }, `${markerKey}_${x}`);
  });
};
AnnotationsMarkersDisplay["displayName"] = "AnnotationsMarkersDisplay";
export {
  AnnotationsMarkersDisplay
};
