// @dynatrace/strato-components/esm/charts/core/components/annotations/components/canvas/displays/AnnotationsTrackDisplay.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { Fragment as Fragment2, useContext } from "react";
import { AnnotationsMarkersDisplay } from "./AnnotationsMarkersDisplay.js";
import { getTokenComputedStyle } from "../../../../../utils/getTokenComputedStyle.js";
import {
  EMPHASIZED_TRACK_SEPARATOR_COLOR,
  LABEL_OFFSET,
  MARKER_LABEL_PADDING,
  TRACK_SEPARATOR_COLOR
} from "../../../constants.js";
import { AnnotationsConfigContext } from "../../../contexts/AnnotationsConfig.context.js";
import { getTrackSeparatorOffset } from "../../../utils/get-track-separator-offset.js";
import { TrackDisplayContext } from "../context/track-display.context.js";
import { AnnotationsLabelRenderer } from "../renderers/AnnotationsLabelRenderer.js";
import { AnnotationsLineRenderer } from "../renderers/AnnotationsLineRenderer.js";
const getLabelYPosition = (index) => getTrackSeparatorOffset(index) + LABEL_OFFSET;
const AnnotationsTrackDisplay = (props) => {
  const { drawTopLine, label, markers, trackIndex, emphasizeBottomLine } = props;
  const { labelsWidth, trackWidth, showLabels } = useContext(
    AnnotationsConfigContext
  );
  const trackFinalXPosition = labelsWidth + trackWidth;
  const topLineY = getTrackSeparatorOffset(trackIndex);
  const bottomLineY = getTrackSeparatorOffset(trackIndex + 1);
  const bottomLineColor = emphasizeBottomLine ? EMPHASIZED_TRACK_SEPARATOR_COLOR : TRACK_SEPARATOR_COLOR;
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    drawTopLine && /* @__PURE__ */ jsx(
      AnnotationsLineRenderer,
      {
        source: { x: labelsWidth, y: topLineY },
        destination: { x: trackFinalXPosition, y: topLineY },
        strokeColor: getTokenComputedStyle(TRACK_SEPARATOR_COLOR)
      }
    ),
    /* @__PURE__ */ jsxs(Fragment2, { children: [
      showLabels && label && /* @__PURE__ */ jsx(
        AnnotationsLabelRenderer,
        {
          text: label,
          x: labelsWidth - MARKER_LABEL_PADDING,
          y: getLabelYPosition(trackIndex),
          maxWidth: labelsWidth - MARKER_LABEL_PADDING,
          ellipsis: true
        }
      ),
      /* @__PURE__ */ jsx(
        TrackDisplayContext.Provider,
        {
          value: {
            offsetX: labelsWidth,
            trackIndex
          },
          children: /* @__PURE__ */ jsx(AnnotationsMarkersDisplay, { markers })
        }
      ),
      /* @__PURE__ */ jsx(
        AnnotationsLineRenderer,
        {
          source: { x: labelsWidth, y: bottomLineY },
          destination: {
            x: trackFinalXPosition,
            y: bottomLineY
          },
          strokeColor: getTokenComputedStyle(bottomLineColor)
        }
      )
    ] })
  ] });
};
AnnotationsTrackDisplay["displayName"] = "AnnotationsTrackDisplay";
export {
  AnnotationsTrackDisplay
};
