// @dynatrace/strato-components/forms/select/components/components.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var components_exports = {};
__export(components_exports, {
  components: () => components
});
module.exports = __toCommonJS(components_exports);
var import_react_select = require("react-select");
var import_Control = require("./Control.js");
var import_Group = require("./Group.js");
var import_GroupHeading = require("./GroupHeading.js");
var import_Menu = require("./Menu.js");
var import_MenuList = require("./MenuList.js");
var import_MultiValue = require("./MultiValue.js");
var import_NoOptionsMessage = require("./NoOptionsMessage.js");
var import_Option = require("./Option.js");
var import_Placeholder = require("./Placeholder.js");
var import_ValueContainer = require("./ValueContainer.js");
const components = {
  Control: import_Control.Control,
  ClearIndicator: import_react_select.components.ClearIndicator,
  DropdownIndicator: import_react_select.components.DropdownIndicator,
  DownChevron: import_react_select.components.DownChevron,
  CrossIcon: import_react_select.components.CrossIcon,
  // Needs a cast here, because in the MultiValue we extend the generic for our _BaseOption and _BaseGroup data types
  // which in turn no longer overlaps with the default generic here.
  Group: import_Group.Group,
  // Needs a cast here, because in the MultiValue we extend the generic for our _BaseOption and _BaseGroup data types
  // which in turn no longer overlaps with the default generic here.
  GroupHeading: import_GroupHeading.GroupHeading,
  IndicatorsContainer: import_react_select.components.IndicatorsContainer,
  IndicatorSeparator: import_react_select.components.IndicatorSeparator,
  Input: import_react_select.components.Input,
  LoadingIndicator: import_react_select.components.LoadingIndicator,
  Menu: import_Menu.Menu,
  // Needs a cast here, because in the MultiValue we extend the generic for our _BaseOption and _BaseGroup data types
  // which in turn no longer overlaps with the default generic here.
  MenuList: import_MenuList.MenuList,
  MenuPortal: import_react_select.components.MenuPortal,
  LoadingMessage: import_react_select.components.LoadingMessage,
  // Needs a cast here, because the types for NoOptionsMessage do expect a defaultProps member to be there
  // coming in from their propTypes
  // https://github.com/JedWatson/react-select/blob/a465818786afd4857c9155bef1e9aa2fe3356097/packages/react-select/src/components/Menu.tsx#L517-L519C3
  NoOptionsMessage: import_NoOptionsMessage.NoOptionsMessage,
  // Needs a cast here, because in the MultiValue we extend the generic for our _BaseOption and _BaseGroup data types
  // which in turn no longer overlaps with the default generic here.
  MultiValue: import_MultiValue.MultiValue,
  MultiValueContainer: import_react_select.components.MultiValueContainer,
  MultiValueLabel: import_react_select.components.MultiValueLabel,
  MultiValueRemove: import_react_select.components.MultiValueRemove,
  // Needs a cast here, because in the Option we extend the generic for our _BaseOption and _BaseGroup data types
  // which in turn no longer overlaps with the default generic here.
  Option: import_Option.Option,
  Placeholder: import_Placeholder.Placeholder,
  SelectContainer: import_react_select.components.SelectContainer,
  SingleValue: import_react_select.components.SingleValue,
  ValueContainer: import_ValueContainer.ValueContainer
};
