// @dynatrace/strato-components/forms/select/components/Control.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Control_exports = {};
__export(Control_exports, {
  Control: () => Control
});
module.exports = __toCommonJS(Control_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_HiddenSelect = require("./HiddenSelect.js");
var import_ValueContainer = require("./ValueContainer.js");
var import_Button = require("../../../buttons/button/Button.js");
var import_useFormControlState = require("../../../core/hooks/useFormControlState.js");
var import_useId = require("../../../core/hooks/useId.js");
var import_useMergeRefs = require("../../../core/hooks/useMergeRefs.js");
var import_OverlayStackingProvider = require("../../../core/providers/OverlayStackingProvider.js");
var import_useFocusRing = require("../../../core/styles/useFocusRing.js");
var import_is_visual_refresh_enabled = require("../../../core/utils/is-visual-refresh-enabled.js");
var import_merge_props = require("../../../core/utils/merge-props.js");
var import_Flex = require("../../../layouts/flex/Flex.js");
var import_FormFieldMessageBaseItem = require("../../forms-core/components/form-field-messages/FormFieldMessageBaseItem.js");
var import_useFormFieldMessagesContext = require("../../forms-core/hooks/useFormFieldMessagesContext.js");
var import_use_select_menu_context = require("../hooks/use-select-menu-context.js");
var import_select_menu_state_reducer = require("../state/select-menu-state-reducer.js");
var styles = __toESM(require("../styles/Select.sty.js"));
function getSelectValueContainer(children) {
  const childrenArray = import_react.Children.toArray(children);
  for (const child of childrenArray) {
    if ((0, import_ValueContainer._isValueContainerComponent)(child)) {
      return child;
    }
  }
  throw new Error("Assertion to make sure the container exists.");
}
function Control(props) {
  const { children, isDisabled, innerRef, innerProps, selectProps } = props;
  const ariaLabelledBy = selectProps["aria-labelledby"];
  const {
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    readOnly,
    ariaDisabled,
    ...remainingProps
  } = selectProps.componentProps.trigger ?? {};
  const rootRef = (0, import_useMergeRefs.useMergeRefs)([
    innerRef,
    selectProps.setTriggerRef
  ]);
  const valueContainer = getSelectValueContainer(children);
  const baseId = (0, import_useId.useId)();
  const triggerId = `strato-select-trigger${baseId}`;
  const customTriggerId = selectProps.customTrigger?.props?.id;
  const { state } = (0, import_useFormFieldMessagesContext.useFormFieldMessagesContext)();
  const [hasFormControlError, hint] = (0, import_useFormControlState.useFormControlState)(
    selectProps.controlState
  );
  const showError = selectProps.controlState ? hasFormControlError : state.showError;
  const isCritical = (0, import_is_visual_refresh_enabled.isVisualRefreshEnabled)() ? showError && !isDisabled : showError;
  const hiddenReactSelectInput = Array.isArray(valueContainer.props.children) ? valueContainer.props.children[1] : null;
  const id = selectProps["selectInputId"];
  const hiddenInput = (0, import_react.useMemo)(
    () => (0, import_react.cloneElement)(hiddenReactSelectInput, {
      ...hiddenReactSelectInput.props,
      key: "hidden-react-select-input",
      id,
      autoComplete: "off",
      "aria-haspopup": "listbox",
      "aria-invalid": showError,
      "aria-labelledby": ariaLabelledBy ?? customTriggerId ?? triggerId,
      "aria-label": selectProps["aria-label"],
      "aria-describedby": selectProps["aria-describedby"],
      "aria-details": selectProps["aria-details"]
    }),
    [
      hiddenReactSelectInput,
      id,
      showError,
      ariaLabelledBy,
      customTriggerId,
      triggerId,
      selectProps
    ]
  );
  const { focusProps, focusClassName } = (0, import_useFocusRing.useFocusRing)({
    color: showError ? "critical" : "neutral",
    focusWithin: true,
    disabled: isDisabled
  });
  const { dispatch } = (0, import_use_select_menu_context.useSelectMenuContext)();
  const handledByTouchRef = (0, import_react.useRef)(false);
  const interactionProps = (0, import_react.useMemo)(() => {
    return {
      onTouchEnd: () => {
        handledByTouchRef.current = true;
        dispatch({
          type: import_select_menu_state_reducer.SelectActions.SELECTFOCUS
        });
        requestAnimationFrame(() => {
          selectProps.filterRef?.current?.inputRef?.focus({
            preventScroll: true
          });
        });
      },
      onClick: () => {
        if (handledByTouchRef.current) {
          handledByTouchRef.current = false;
          return;
        }
        dispatch({
          type: import_select_menu_state_reducer.SelectActions.SELECTFOCUS
        });
      },
      onKeyDown: (event) => {
        if ((event.key === "Enter" || event.key === " ") && !selectProps.isMulti) {
          dispatch({
            type: import_select_menu_state_reducer.SelectActions.CLOSE
          });
        }
      }
    };
  }, [dispatch, selectProps.isMulti]);
  const customTrigger = (0, import_react.useMemo)(() => {
    if (selectProps.customTrigger) {
      return (0, import_react.cloneElement)(
        selectProps.customTrigger,
        {
          key: "custom-trigger",
          ref: rootRef,
          id: triggerId,
          ...(0, import_merge_props.mergeProps)(
            selectProps.customTrigger.props,
            innerProps,
            ...isDisabled ? [] : [interactionProps]
          )
        },
        [selectProps.customTrigger.props?.children, hiddenInput]
      );
    }
  }, [
    hiddenInput,
    innerProps,
    interactionProps,
    rootRef,
    selectProps.customTrigger,
    triggerId
  ]);
  if (customTrigger) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayStackingProvider.OverlayStackingProvider, { children: customTrigger });
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { padding: 0, flexDirection: "column", gap: 4, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayStackingProvider.OverlayStackingProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      import_Button.Button,
      {
        ref: rootRef,
        id: triggerId,
        textAlign: "start",
        "data-testid": dataTestId,
        disabled: isDisabled,
        "aria-disabled": ariaDisabled,
        width: selectProps.triggerWidth,
        ...innerProps,
        variant: "emphasized",
        ...isDisabled ? void 0 : interactionProps,
        color: isCritical ? "critical" : "neutral",
        className: (0, import_clsx.default)(
          "strato-components-select-control-trigger",
          styles.selectTrigger({
            fullWidth: selectProps.triggerWidth === "full" || selectProps.triggerWidth === "100%",
            showError,
            readOnly
          }),
          focusClassName,
          consumerClassName
        ),
        style: consumerStyle,
        as: "div",
        ...focusProps,
        ...remainingProps,
        children: [
          selectProps.prefix && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_Button.Button.Prefix,
            {
              "data-testid": selectProps.prefix.props["data-testid"],
              className: selectProps.prefix.props.className,
              style: selectProps.prefix.props.style,
              children: selectProps.prefix.props.children
            }
          ),
          /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Button.Button.Label, { children: [
            hiddenInput,
            valueContainer
          ] }),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Suffix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "div",
            {
              className: (0, import_clsx.default)(
                styles.selectTriggerIconWrapper({
                  isOverlayOpen: selectProps.menuIsOpen,
                  readOnly
                })
              ),
              children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ChevronDownSmallIcon, {})
            }
          ) })
        ]
      }
    ) }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_HiddenSelect.HiddenSelect,
      {
        ref: selectProps.hiddenSelectRef,
        required: selectProps.isRequired,
        selectedValues: selectProps.value,
        name: selectProps.hiddenSelectName,
        disabled: isDisabled || readOnly,
        controlState: selectProps.controlState
      }
    ),
    hint && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_FormFieldMessageBaseItem.InternalBaseItem,
      {
        variant: showError ? "error" : "hint",
        disabled: isDisabled || ariaDisabled,
        children: hint
      }
    )
  ] });
}
