// @dynatrace/strato-components/forms/select/components/Group.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Group_exports = {};
__export(Group_exports, {
  Group: () => Group
});
module.exports = __toCommonJS(Group_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_select = require("react-select");
var import_sort_selected_options_first = require("./sort-selected-options-first.js");
var import_useBehavioralTrackingProps = require("../../../core/hooks/useBehavioralTrackingProps.js");
function Group(props) {
  const initialRender = (0, import_react.useRef)(true);
  const sortingMap = (0, import_react.useRef)(/* @__PURE__ */ new WeakMap());
  const { children, innerProps, ...remainingProps } = props;
  const {
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId
  } = props.data;
  const [groupTrackingProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(props.data);
  let childrenArray = import_react.Children.toArray(children);
  if (props.selectProps.showSelectedOptionsFirst) {
    if (initialRender.current === true) {
      [childrenArray, sortingMap.current] = (0, import_sort_selected_options_first.sortSelectedOptions)(childrenArray);
    } else {
      const [newChildrenArray, newSortingMap] = (0, import_sort_selected_options_first.applyStoredSelectionSortingOptions)(childrenArray, sortingMap.current);
      childrenArray = newChildrenArray;
      if (newSortingMap) {
        sortingMap.current = newSortingMap;
      }
    }
  }
  initialRender.current = false;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_react_select.components.Group,
    {
      ...remainingProps,
      innerProps: Object.assign(innerProps ?? {}, {
        role: "group",
        className: consumerClassName,
        style: consumerStyle,
        "data-testid": dataTestId,
        ...groupTrackingProps
      }),
      children: childrenArray
    }
  );
}
Group.displayName = "Select.SelectGroup";
