// @dynatrace/strato-components/forms/select/components/Menu.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Menu_exports = {};
__export(Menu_exports, {
  Menu: () => Menu
});
module.exports = __toCommonJS(Menu_exports);
var import_jsx_runtime = (
  // The empty content is need so the TextInput does not consume the required coming from the FormField
  require("react/jsx-runtime")
);
var import_dynamic = require("@vanilla-extract/dynamic");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_dom = require("react-dom");
var import_react_intl = require("react-intl");
var import_react_popper = require("react-popper");
var import_borders = __toESM(require("@dynatrace/strato-design-tokens/borders"));
var import_strato_icons = require("@dynatrace/strato-icons");
var import_Button = require("../../../buttons/button/Button.js");
var import_ProgressCircle = require("../../../content/progress/ProgressCircle.js");
var import_useId = require("../../../core/hooks/useId.js");
var import_useMergeRefs = require("../../../core/hooks/useMergeRefs.js");
var import_useOverlayStacking = require("../../../core/hooks/useOverlayStacking.js");
var import_OverlayStackingProvider = require("../../../core/providers/OverlayStackingProvider.js");
var import_Divider = require("../../../layouts/divider/Divider.js");
var import_Flex = require("../../../layouts/flex/Flex.js");
var import_Surface = require("../../../layouts/surface/Surface.js");
var import_variables_css = require("../../../layouts/surface/variables.sty.js");
var import_useOverlay = require("../../../overlays/hooks/useOverlay.js");
var import_popper_modifiers = require("../../../overlays/utils/popper-modifiers.js");
var import_Text = require("../../../typography/text/Text.js");
var import_FormFieldContext = require("../../forms-core/contexts/FormFieldContext.js");
var import_FormFieldMessageContext = require("../../forms-core/contexts/FormFieldMessageContext.js");
var import_SearchInput = require("../../search-input/SearchInput.js");
var import_use_select_menu_context = require("../hooks/use-select-menu-context.js");
var import_select_menu_state_reducer = require("../state/select-menu-state-reducer.js");
var styles = __toESM(require("../styles/Select.sty.js"));
const sameWidthModifier = {
  name: "sameWidth-as-trigger",
  enabled: true,
  fn: ({ state }) => {
    state.styles.popper.width = `${state.rects.reference.width}px`;
    state.styles.popper.height = "min-content";
  },
  phase: "beforeWrite",
  requires: ["computeStyles"]
};
const heightModifier = {
  name: "select-height-handler",
  enabled: true,
  fn: ({ state }) => {
    const viewportHeight = Math.max(
      document.documentElement.clientHeight || 0,
      window.innerHeight || 0
    );
    const topOffset = state.rects.reference.y;
    const bottomOffset = viewportHeight - (state.rects.reference.y + state.rects.reference.height);
    const offset = state.options.placement.includes("bottom") ? bottomOffset : topOffset;
    if (offset < state.rects.popper.height) {
      const maxHeightBase = viewportHeight / 2 > 320 ? viewportHeight / 2 : 320;
      state.styles.popper.maxHeight = `${maxHeightBase - state.rects.reference.height - 20}px`;
    }
  },
  phase: "beforeWrite",
  requires: ["computeStyles"]
};
function Menu(props) {
  const { children, innerRef, innerProps, selectProps } = props;
  const {
    filterRef,
    clearSelectionRef,
    triggerRef,
    inputValue,
    isMulti,
    onInputChange
  } = selectProps;
  const {
    className: contentConsumerClassName,
    style: contentConsumerStyle,
    "data-testid": contentDataTestId,
    ...remainingContentProps
  } = selectProps.componentProps.content ?? {};
  const {
    className: filterConsumerClassName,
    style: filterConsumerStyle,
    "data-testid": filterDataTestId,
    ...remainingFilterProps
  } = selectProps.componentProps.filter ?? {};
  const [portalRef, setPortalRef] = (0, import_react.useState)(null);
  const [menuWidth, setMenuWidth] = (0, import_react.useState)(
    triggerRef?.offsetWidth
  );
  const rootPortalRef = (0, import_useMergeRefs.useMergeRefs)([innerRef, setPortalRef]);
  const stackingContext = (0, import_useOverlayStacking.useOverlayStacking)(void 0, "overlay");
  const { styles: positioningStyles } = (0, import_react_popper.usePopper)(triggerRef, portalRef, {
    placement: "bottom-start",
    modifiers: [
      {
        name: "offset",
        options: {
          offset: [0, 4]
        }
      },
      sameWidthModifier,
      heightModifier,
      import_popper_modifiers.observeOverlaySize
    ]
  });
  const { dispatch } = (0, import_use_select_menu_context.useSelectMenuContext)();
  const clearFilterValue = (0, import_react.useCallback)(() => {
    if (inputValue !== "") {
      onInputChange("", {
        action: "input-change",
        prevInputValue: inputValue
      });
    }
  }, [inputValue, onInputChange]);
  (0, import_useOverlay.useOverlay)(
    {
      onClose: () => {
        dispatch({
          type: import_select_menu_state_reducer.SelectActions.CLOSE
        });
        clearFilterValue();
        setMenuWidth(triggerRef?.offsetWidth);
      },
      isOpen: selectProps.menuIsOpen,
      isDismissible: true,
      shouldCloseOnInteractOutside: (element) => {
        return !selectProps.triggerRef?.contains(element);
      }
    },
    innerRef
  );
  (0, import_react.useEffect)(() => {
    if (selectProps.menuIsOpen) {
      filterRef?.current?.inputRef?.focus({ preventScroll: true });
    }
  }, [filterRef, selectProps.menuIsOpen]);
  const hiddenTrigger = (0, import_react.useMemo)(
    () => selectProps.triggerRef?.querySelector("input"),
    [selectProps.triggerRef]
  );
  (0, import_react.useEffect)(() => {
    return () => {
      if (!document.activeElement || document.activeElement === document.body) {
        hiddenTrigger?.focus();
      }
    };
  }, [hiddenTrigger]);
  const intl = (0, import_react_intl.useIntl)();
  const ariaLabelId = (0, import_useId.useId)();
  const allSelected = (0, import_react.useMemo)(() => {
    return isMulti && selectProps.allOptions?.length === selectProps.value?.length;
  }, [isMulti, selectProps.allOptions, selectProps.value]);
  const handleSelectAllInteraction = () => {
    if (!selectProps.loading) {
      if (allSelected) {
        selectProps.onChange([], {
          action: "clear",
          removedValues: selectProps.value
        });
      } else {
        selectProps.onChange(
          selectProps.allOptions,
          {}
        );
      }
    }
  };
  const dataDtProperties = `state:${allSelected ? "deselect" : "select"}`;
  const noOptionsAvailable = selectProps.allOptions === void 0 || selectProps.allOptions.length === 0;
  const isFiltering = selectProps.customFilter && inputValue !== void 0 && inputValue !== "";
  const showClearSelection = selectProps.clearable && selectProps.value && isMulti === false && !isFiltering;
  const showSelectAll = selectProps.clearable && isMulti && !isFiltering && !noOptionsAvailable;
  return (0, import_react_dom.createPortal)(
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Surface.Surface,
      {
        "data-testid": contentDataTestId,
        ref: rootPortalRef,
        ...innerProps,
        elevation: "floating",
        padding: 4,
        className: (0, import_clsx.default)(
          styles.selectMenuOverlay({
            isTriggerWidth: selectProps.contentWidth === "trigger" || selectProps.customFilter && selectProps.contentWidth === "content"
          }),
          contentConsumerClassName
        ),
        style: {
          zIndex: stackingContext.style.zIndex,
          ...positioningStyles.popper,
          width: selectProps.contentWidth !== "trigger" ? selectProps.contentWidth : menuWidth,
          ...contentConsumerStyle,
          ...(0, import_dynamic.assignInlineVars)({
            [import_variables_css.surfaceBorderRadius]: import_borders.default.Radius.Surface.Subdued
          })
        },
        onMouseDown: (e) => e.stopPropagation(),
        ...remainingContentProps,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_OverlayStackingProvider.OverlayStackingProvider, { children: [
          selectProps.customFilter && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_FormFieldContext._FormFieldContext.Provider, { value: {}, children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_FormFieldMessageContext.FormFieldMessageContext.Provider,
              {
                value: {
                  state: {
                    customMessages: [],
                    formControlMessages: [],
                    showError: false
                  }
                },
                children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  import_SearchInput.SearchInput,
                  {
                    "data-testid": filterDataTestId,
                    ref: filterRef,
                    variant: "minimal",
                    "aria-label": intl.formatMessage({
                      defaultMessage: "Search available options",
                      id: "4PipZ2Z07Q9epib1",
                      description: "Aria label for the filter inside the select"
                    }),
                    placeholder: intl.formatMessage({
                      defaultMessage: "Search",
                      id: "CerADEHqR/6IVmHr",
                      description: "Placeholder for the filter inside the select"
                    }),
                    value: inputValue,
                    onChange: (value) => {
                      onInputChange(value, {
                        action: "input-change",
                        prevInputValue: inputValue
                      });
                    },
                    className: (0, import_clsx.default)(styles.filter, filterConsumerClassName),
                    style: filterConsumerStyle,
                    ...remainingFilterProps
                  }
                )
              }
            ),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_Divider.Divider,
              {
                className: (0, import_clsx.default)(
                  styles.selectMenuDivider,
                  "strato-select-divider"
                )
              }
            )
          ] }),
          showClearSelection && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_Button.Button,
              {
                ref: clearSelectionRef,
                width: "full",
                textAlign: "start",
                onClick: () => {
                  selectProps.onChange(
                    null,
                    {
                      action: "clear",
                      removedValues: selectProps.value
                    }
                  );
                  if (isMulti === false) {
                    dispatch({
                      type: import_select_menu_state_reducer.SelectActions.CLOSE
                    });
                    clearFilterValue();
                  }
                },
                onKeyDown: (event) => {
                  if (event.key === " " || event.key === "Enter") {
                    selectProps.onChange(
                      null,
                      {
                        action: "clear",
                        removedValues: selectProps.value
                      }
                    );
                    event.preventDefault();
                    event.stopPropagation();
                    if (isMulti === false) {
                      dispatch({
                        type: import_select_menu_state_reducer.SelectActions.CLOSE
                      });
                      clearFilterValue();
                    }
                    if (filterRef?.current && isMulti) {
                      filterRef.current.inputRef?.focus();
                    } else {
                      const hiddentrigger = triggerRef?.querySelector("input");
                      hiddentrigger?.focus();
                    }
                  }
                },
                children: intl.formatMessage({
                  defaultMessage: "Clear selection",
                  id: "/H3rbOvaR6bAasVo",
                  description: "Select component clear selection button."
                })
              }
            ),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_Divider.Divider,
              {
                className: (0, import_clsx.default)(
                  styles.selectMenuDivider,
                  "strato-select-divider"
                )
              }
            )
          ] }),
          showSelectAll && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
              import_Flex.Flex,
              {
                tabIndex: 0,
                role: "option",
                className: (0, import_clsx.default)(
                  "strato-select-option",
                  styles.selectOptionMulti,
                  styles.selectAll,
                  {
                    "strato-select-option--selected": selectProps.value?.length > 0,
                    "strato-select-option--disabled": selectProps.loading
                  }
                ),
                ref: selectProps.clearSelectionRef,
                onClick: () => {
                  handleSelectAllInteraction();
                },
                onKeyDown: (event) => {
                  if (event.key === " " || event.key === "Enter") {
                    handleSelectAllInteraction();
                    event.preventDefault();
                    event.stopPropagation();
                  }
                },
                "data-dt-name": "SelectDeselectAll",
                "data-dt-properties": dataDtProperties,
                children: [
                  /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
                    "div",
                    {
                      className: (0, import_clsx.default)(
                        styles.selectOptionMultiIndicator({
                          selected: selectProps.value?.length > 0
                        })
                      ),
                      children: [
                        allSelected ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                          import_strato_icons.CheckmarkSmallIcon,
                          {
                            className: styles.selectOptionMultiIndicatorIcon
                          }
                        ) : null,
                        !allSelected && selectProps.value?.length > 0 ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                          import_strato_icons.MinusIcon,
                          {
                            className: styles.selectOptionMultiIndicatorIcon
                          }
                        ) : null
                      ]
                    }
                  ),
                  intl.formatMessage({
                    defaultMessage: "Select all",
                    id: "EsLvnx9UfUL/bZ/r",
                    description: "Select all option in the Select component."
                  })
                ]
              }
            ),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_Divider.Divider,
              {
                className: (0, import_clsx.default)(
                  styles.selectMenuDivider,
                  "strato-select-divider"
                )
              }
            )
          ] }),
          selectProps.loading ? /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
            import_Flex.Flex,
            {
              paddingY: 6,
              paddingLeft: 8,
              paddingRight: 12,
              className: styles.selectMenuLoadingWrapper,
              children: [
                /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ProgressCircle.ProgressCircle, { size: "small", "aria-labelledby": ariaLabelId }),
                /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { id: ariaLabelId, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  import_react_intl.FormattedMessage,
                  {
                    description: "Label for the loading state.",
                    defaultMessage: "Loading",
                    id: "cycz+bYG+VH7Of4e"
                  }
                ) })
              ]
            }
          ) : children
        ] })
      }
    ),
    document.body
  );
}
