// @dynatrace/strato-components/forms/select/components/MenuList.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MenuList_exports = {};
__export(MenuList_exports, {
  MenuList: () => MenuList
});
module.exports = __toCommonJS(MenuList_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_virtual = require("@tanstack/react-virtual");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_NoOptionsMessage = require("./NoOptionsMessage.js");
var import_sort_selected_options_first = require("./sort-selected-options-first.js");
var import_useMergeRefs = require("../../../core/hooks/useMergeRefs.js");
var styles = __toESM(require("../styles/Select.sty.js"));
function MenuList(props) {
  const { children, selectProps, innerProps } = props;
  const parentRef = (0, import_react.useRef)(null);
  const [notifyParentRefChange, setNotifyParentRefChange] = (0, import_react.useState)();
  const initialRender = (0, import_react.useRef)(true);
  const sortingMap = (0, import_react.useRef)(/* @__PURE__ */ new WeakMap());
  let childrenArray = import_react.Children.toArray(children);
  const currentChildrenLength = (0, import_react.useRef)(childrenArray.length);
  if (selectProps.showSelectedOptionsFirst && !(childrenArray.length === 1 && childrenArray[0].type === import_NoOptionsMessage.NoOptionsMessage)) {
    if (initialRender.current === true) {
      [childrenArray, sortingMap.current] = (0, import_sort_selected_options_first.sortSelectedOptions)(childrenArray);
    } else if (currentChildrenLength.current !== childrenArray.length) {
      currentChildrenLength.current = childrenArray.length;
      [childrenArray, sortingMap.current] = (0, import_sort_selected_options_first.sortSelectedOptions)(childrenArray);
    } else {
      const [newChildrenArray, newSortingMap] = (0, import_sort_selected_options_first.applyStoredSelectionSortingOptions)(childrenArray, sortingMap.current);
      childrenArray = newChildrenArray;
      if (newSortingMap) {
        sortingMap.current = newSortingMap;
      }
    }
  }
  initialRender.current = false;
  const optionVirtualizer = (0, import_react_virtual.useVirtualizer)({
    count: childrenArray.length,
    getScrollElement: () => parentRef.current,
    overscan: 5,
    estimateSize: () => 32
  });
  const menulistref = (0, import_useMergeRefs.useMergeRefs)([
    props.innerRef,
    parentRef,
    setNotifyParentRefChange,
    selectProps.contentRef
  ]);
  (0, import_react.useImperativeHandle)(
    selectProps.menuListFunctionRef,
    () => ({
      scrollToBottom: () => {
        optionVirtualizer.scrollToIndex(childrenArray.length);
      }
    })
  );
  (0, import_react.useEffect)(() => {
    requestAnimationFrame(() => {
      optionVirtualizer.measure();
    });
  }, [notifyParentRefChange, childrenArray.length]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      ...innerProps,
      ref: menulistref,
      className: (0, import_clsx.default)(styles.selectMenuListWrapper, "strato-select-menulist"),
      style: {
        width: optionVirtualizer.scrollOffset !== null && optionVirtualizer.scrollOffset > 0 && selectProps.contentWidth === "max-content" ? optionVirtualizer.scrollRect?.width : void 0
      },
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        "div",
        {
          role: "presentation",
          style: {
            height: optionVirtualizer.getTotalSize(),
            display: "grid",
            gridTemplateRows: optionVirtualizer.measurementsCache.map((el) => `${el.size}px`).join(" ")
          },
          className: styles.selectVirtualizationWrapper,
          children: optionVirtualizer.getVirtualItems().map((virtualRow) => {
            const child = childrenArray?.[virtualRow.index];
            return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              "div",
              {
                className: "strato-select-option",
                "data-index": virtualRow.index,
                style: {
                  gridRow: virtualRow.index + 1,
                  minWidth: 0,
                  height: "fit-content"
                },
                ref: (el) => {
                  optionVirtualizer.measureElement(el);
                },
                children: child
              },
              virtualRow.key
            );
          })
        },
        `wrapper-${selectProps.options.length}`
      )
    }
  );
}
