// @dynatrace/strato-components/forms/select/components/NoOptionsMessage.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var NoOptionsMessage_exports = {};
__export(NoOptionsMessage_exports, {
  NoOptionsMessage: () => NoOptionsMessage
});
module.exports = __toCommonJS(NoOptionsMessage_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react_intl = require("react-intl");
var import_can_receive_default_text_format = require("../../../core/utils/can-receive-default-text-format.js");
var import_Text = require("../../../typography/text/Text.js");
var styles = __toESM(require("../styles/Select.sty.js"));
function NoOptionsMessage(props) {
  const selectEmptyState = props.selectProps.emptyState?.props.children;
  const {
    className: consumerClassName,
    style: consumerStyle,
    ...remainingProps
  } = props.selectProps.componentProps.emptyState ?? {};
  const intl = (0, import_react_intl.useIntl)();
  if (!selectEmptyState || (0, import_can_receive_default_text_format.canReceiveDefaultTextFormat)(selectEmptyState)) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Text.Text,
      {
        textStyle: "base",
        style: consumerStyle,
        className: (0, import_clsx.default)(styles.selectNoOptionsMessage, consumerClassName),
        as: "p",
        ...remainingProps,
        children: selectEmptyState ?? intl.formatMessage({
          defaultMessage: "No results found",
          id: "U09aNQxYT1yrJ4eJ",
          description: "Default message when no options are available in the select."
        })
      }
    );
  } else {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Text.Text,
      {
        textStyle: "base",
        style: consumerStyle,
        className: (0, import_clsx.default)(styles.selectNoOptionsMessage, consumerClassName),
        as: "p",
        ...remainingProps,
        children: selectEmptyState
      }
    );
  }
}
