// @dynatrace/strato-components/forms/select/components/Option.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Option_exports = {};
__export(Option_exports, {
  Option: () => Option
});
module.exports = __toCommonJS(Option_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_strato_icons = require("@dynatrace/strato-icons");
var import_TextEllipsisAroundChildren = require("../../../core/components/TextEllipsisAroundChildren.js");
var import_useBehavioralTrackingProps = require("../../../core/hooks/useBehavioralTrackingProps.js");
var import_getVirtualFocusStyles = require("../../../core/styles/getVirtualFocusStyles.js");
var import_Flex = require("../../../layouts/flex/Flex.js");
var import_use_select_menu_context = require("../hooks/use-select-menu-context.js");
var import_select_menu_state_reducer = require("../state/select-menu-state-reducer.js");
var styles = __toESM(require("../styles/Select.sty.js"));
function Option(props) {
  const {
    children,
    isSelected,
    innerProps,
    isFocused,
    isDisabled,
    isMulti,
    selectProps
  } = props;
  const {
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "aria-label": ariaLabel,
    prefix,
    suffix
  } = props.data;
  const [behavioralTrackingProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(props.data);
  const [prefixTrackingProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(
    prefix.componentProps ?? {}
  );
  const [suffixTrackingProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(
    suffix.componentProps ?? {}
  );
  const { dispatch } = (0, import_use_select_menu_context.useSelectMenuContext)();
  const isVirtuallyFocused = selectProps.focusOnOption && isFocused;
  const { className: virtualFocusClassName } = (0, import_getVirtualFocusStyles.getVirtualFocusStyles)(
    isVirtuallyFocused,
    isSelected && !isMulti ? "primary" : "neutral"
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_Flex.Flex,
    {
      "data-testid": dataTestId,
      role: "option",
      ...innerProps,
      className: (0, import_clsx.default)(
        "strato-select-option",
        isMulti === true ? styles.selectOptionMulti : styles.selectOptionSingle,
        virtualFocusClassName,
        {
          "strato-select-option--selected": isSelected,
          "strato-select-option--disabled": isDisabled
        },
        consumerClassName
      ),
      style: consumerStyle,
      onClick: (event) => {
        if (!isDisabled) {
          if (isMulti === false) {
            dispatch({
              type: import_select_menu_state_reducer.SelectActions.CLOSE
            });
          }
          innerProps.onClick?.(event);
        }
      },
      "aria-selected": isSelected,
      "aria-label": ariaLabel,
      ...behavioralTrackingProps,
      children: [
        isMulti === true ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          "div",
          {
            className: (0, import_clsx.default)(
              styles.selectOptionMultiIndicator({
                selected: isSelected
              })
            ),
            children: isSelected === true ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_strato_icons.CheckmarkSmallIcon,
              {
                className: styles.selectOptionMultiIndicatorIcon
              }
            ) : null
          }
        ) : null,
        props.selectProps.optionsHavePrefix ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          "span",
          {
            className: (0, import_clsx.default)(
              styles.selectOptionPrefix,
              prefix.componentProps?.className
            ),
            style: prefix.componentProps?.style,
            "data-testid": prefix.componentProps?.["data-testid"],
            ...prefixTrackingProps,
            children: prefix.component
          }
        ) : null,
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TextEllipsisAroundChildren.TextEllipsisAroundChildren, { children }),
        suffix.component ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          "span",
          {
            className: (0, import_clsx.default)(
              styles.selectOptionSuffix,
              suffix.componentProps?.className
            ),
            style: suffix.componentProps?.style,
            "data-testid": suffix.componentProps?.["data-testid"],
            ...suffixTrackingProps,
            children: suffix.component
          }
        ) : null
      ]
    }
  );
}
