// @dynatrace/strato-components/forms/select/components/Placeholder.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Placeholder_exports = {};
__export(Placeholder_exports, {
  Placeholder: () => Placeholder
});
module.exports = __toCommonJS(Placeholder_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react_intl = require("react-intl");
var import_react_select = require("react-select");
var import_can_receive_default_text_format = require("../../../core/utils/can-receive-default-text-format.js");
var styles = __toESM(require("../styles/Select.sty.js"));
function Placeholder(props) {
  const { selectProps } = props;
  const intl = (0, import_react_intl.useIntl)();
  if (selectProps.placeholderText && !(0, import_can_receive_default_text_format.canReceiveDefaultTextFormat)(selectProps.placeholderText)) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_select.components.Placeholder, { ...props, className: styles.placeholderColor, children: selectProps.placeholderText });
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_react_select.components.Placeholder,
    {
      ...props,
      className: (0, import_clsx.default)(
        styles.placeholderColor,
        styles.selectValueContainerTruncationBox
      ),
      children: selectProps.placeholderText ?? (selectProps.isMulti ? intl.formatMessage({
        defaultMessage: "Select options",
        id: "3ZX0IG+5tekB89nr",
        description: "Default placeholder text for the multi select."
      }) : intl.formatMessage({
        defaultMessage: "Select an option",
        id: "ggfg/kIzwBCNyGCm",
        description: "Default placeholder text for the single select."
      }))
    }
  );
}
