// @dynatrace/strato-components/esm/charts/core/components/selection-area/components/Knob.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { useGesture } from "@use-gesture/react";
import clsx from "clsx";
import {
  KNOB_HEIGHT,
  KNOB_WIDTH,
  SELECTED_AREA_RADIUS
} from "../../overlays/constants.js";
import { knob, transparentDraggingArea } from "../SelectionArea.sty.js";
const Knob = ({
  y,
  x,
  interactable,
  onKeyDown,
  focusable,
  onDragMove,
  onDragStart,
  onDragEnd,
  edgeId,
  hoveredEdge,
  pointerHandlers
}) => {
  const handlers = useGesture(
    {
      ...pointerHandlers,
      onDragStart,
      onDragEnd,
      onDrag: ({ xy }) => {
        onDragMove(xy[0]);
      }
    },
    {
      drag: {
        threshold: 2
      }
    }
  );
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(
      "rect",
      {
        y: y - KNOB_HEIGHT / 2,
        x: x - KNOB_WIDTH / 2 * 2,
        className: clsx(transparentDraggingArea),
        style: {
          pointerEvents: interactable ? "all" : "none"
        },
        ...handlers()
      }
    ),
    /* @__PURE__ */ jsx(
      "rect",
      {
        y,
        x,
        rx: SELECTED_AREA_RADIUS,
        className: clsx(
          knob,
          "knob",
          hoveredEdge === edgeId && "knob--hovered"
        ),
        tabIndex: focusable ? 0 : -1,
        onKeyDown,
        "aria-label": edgeId
      }
    )
  ] });
};
export {
  Knob
};
