// @dynatrace/strato-components/esm/charts/core/components/selection-area/components/SelectionEdge.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { isUndefined } from "lodash-es";
import Colors from "@dynatrace/strato-design-tokens/colors";
import { Knob } from "./Knob.js";
import { usePoliteContainer } from "../../../../../overlays/hooks/usePoliteContainer.js";
import { BottomMagnifier } from "../../crosshair/components/BottomMagnifier.js";
import { TopMagnifier } from "../../crosshair/components/TopMagnifier.js";
import { KNOB_WIDTH } from "../../overlays/constants.js";
const SelectionEdge = ({
  x,
  y,
  focusable,
  interactable,
  onKeyDown,
  onDragMove,
  onDragStart,
  onDragEnd,
  topLabel,
  bottomLabel,
  chartBounds,
  isBottomLabelRotated = false,
  isTopLabelRotated = false,
  edgeId,
  hoveredEdge,
  pointerHandlers
}) => {
  const Portal = usePoliteContainer();
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    !isUndefined(bottomLabel) ? /* @__PURE__ */ jsx(Portal, { children: /* @__PURE__ */ jsx(
      BottomMagnifier,
      {
        label: bottomLabel,
        mouseX: x,
        offset: 0,
        color: Colors.Border.Primary.Accent,
        chartBounds,
        isRotated: isBottomLabelRotated
      }
    ) }) : null,
    !isUndefined(topLabel) ? /* @__PURE__ */ jsx(Portal, { children: /* @__PURE__ */ jsx(
      TopMagnifier,
      {
        label: topLabel,
        mouseX: x,
        offset: 0,
        color: Colors.Border.Primary.Accent,
        chartBounds,
        isRotated: isTopLabelRotated
      }
    ) }) : null,
    /* @__PURE__ */ jsx(
      Knob,
      {
        edgeId,
        hoveredEdge,
        y,
        x: x - KNOB_WIDTH / 2,
        interactable,
        focusable,
        onKeyDown,
        onDragMove,
        onDragStart,
        onDragEnd,
        pointerHandlers
      }
    )
  ] });
};
export {
  SelectionEdge
};
