// @dynatrace/strato-components/esm/charts/core/components/threshold-bar/hooks/useThresholdBarTooltip.js
import {
  useEffect,
  useMemo,
  useState
} from "react";
import { useIntl } from "react-intl";
import { THRESHOLD_MESSAGES } from "../messages/messages.js";
import { groupOverlappingMarkers } from "../utils/group-overlapping-markers.js";
const TOOLTIP_Y_OFFSET = 22;
function useThresholdBarTooltip(markers, onHover) {
  const groupedMarkers = useMemo(
    () => groupOverlappingMarkers(markers),
    [markers]
  );
  const intl = useIntl();
  const [hoveredIndex, setHoveredIndex] = useState();
  const [tooltipPosition, setTooltipPosition] = useState();
  const [tooltipItems, setTooltipItems] = useState([]);
  useEffect(() => {
    if (hoveredIndex === void 0) {
      return;
    }
    const onScroll = () => setHoveredIndex(void 0);
    window.addEventListener("scroll", onScroll, { once: true, capture: true });
    return () => window.removeEventListener("scroll", onScroll, { capture: true });
  }, [hoveredIndex]);
  function handleMouseMove(event, index) {
    setHoveredIndex(index);
    const tooltipPosition2 = {
      x: event.clientX,
      y: event.clientY + TOOLTIP_Y_OFFSET
    };
    const marker = markers[index];
    const belongingGroup = groupedMarkers.find(
      (group) => group.some((item) => {
        return item.y1 === marker.y1 && item.y0 === marker.y0;
      })
    );
    if (belongingGroup) {
      const tooltipItems2 = belongingGroup.filter((item) => !item.isShadow).map((item, index2) => ({
        id: item.id,
        color: { Default: item.color },
        values: [item.value],
        name: item.label ?? intl.formatMessage(THRESHOLD_MESSAGES.thresholdLabel)
      }));
      setTooltipPosition(tooltipPosition2);
      setTooltipItems(tooltipItems2);
      onHover?.(marker.id);
    }
  }
  function handleMouseLeave() {
    setHoveredIndex(void 0);
    setTooltipItems([]);
    onHover?.(void 0);
  }
  return {
    showTooltip: hoveredIndex !== void 0 && tooltipPosition,
    tooltipPosition,
    tooltipItems,
    handleMouseMove,
    handleMouseLeave
  };
}
export {
  useThresholdBarTooltip
};
