// @dynatrace/strato-components/esm/charts/xy-chart/processing-strategies/utils/access-datapoint/get-bar-datapoints.js
import { getAxisType, getDatumValue, getPoint } from "./get-datapoints.js";
import { numberOfCoordinatesByDimension } from "./number-of-coordinates-by-dimension.js";
function resolveLabel(datum, labelAccessor) {
  if (!labelAccessor) {
    return void 0;
  }
  const raw = getDatumValue(datum, labelAccessor);
  if (raw === void 0 || raw === null) {
    return void 0;
  }
  const label = String(raw);
  return label.length > 0 ? label : void 0;
}
function getBarDatapoint(datum, seriesProps, xAxis, yAxis) {
  const xAxisType = getAxisType("x", xAxis, yAxis);
  const yAxisType = getAxisType("y", xAxis, yAxis);
  const point = {
    x0: getPoint(datum, seriesProps.x0Accessor, xAxisType),
    x1: getPoint(datum, seriesProps.x1Accessor, xAxisType),
    y0: getPoint(datum, seriesProps.y0Accessor, yAxisType),
    y1: getPoint(datum, seriesProps.y1Accessor, yAxisType)
  };
  const { xCoordinateCount, yCoordinateCount } = numberOfCoordinatesByDimension(point);
  if (xCoordinateCount === 0 || yCoordinateCount === 0 || xCoordinateCount + yCoordinateCount === 1) {
    return void 0;
  }
  if (xCoordinateCount + yCoordinateCount === 2) {
    const invalidCombinations = point.x0 !== void 0 && point.y0 !== void 0 || point.x1 !== void 0 && point.y1 !== void 0;
    if (invalidCombinations) {
      return void 0;
    }
  }
  return {
    ...datum,
    ...point,
    label: resolveLabel(datum, seriesProps.labelAccessor)
  };
}
export {
  getBarDatapoint
};
