// @dynatrace/strato-components/esm/charts/xy-chart/processing-strategies/utils/access-datapoint/get-datapoints.js
import { get, isFunction, isUndefined } from "lodash-es";
import { XYAxisTypeTime } from "../../../types/axis.js";
import { isCategorical, isTime } from "../../../types/guards.js";
import { castDataType } from "../cast-data/cast-data-types.js";
function getDatumValue(datum, path) {
  if (!path) {
    return void 0;
  }
  if (isFunction(path)) {
    return path(datum);
  }
  return get(datum, path);
}
function getPoint(datum, path, axisType) {
  const value = getDatumValue(datum, path);
  return castDataType(value, axisType);
}
function getBaseDatapoint(datum, seriesProps, xAxis, yAxis) {
  const x0 = getPoint(datum, seriesProps.x0Accessor, xAxis.type);
  const y0 = getPoint(datum, seriesProps.y0Accessor, yAxis.type);
  let x1;
  let y1;
  if (isUndefined(x0) || isUndefined(y0)) {
    return;
  }
  if (isTime(xAxis.type) && seriesProps.x1Accessor) {
    x1 = getPoint(datum, seriesProps.x1Accessor, XYAxisTypeTime);
  }
  if (isTime(yAxis.type) && seriesProps.y1Accessor) {
    y1 = getPoint(datum, seriesProps.y1Accessor, XYAxisTypeTime);
  }
  return { ...datum, x0, x1, y0, y1 };
}
function getDotDatapoint(datum, seriesProps, xAxis, yAxis) {
  return getBaseDatapoint(datum, seriesProps, xAxis, yAxis);
}
function getRectDatapoint(datum, seriesProps, xAxis, yAxis) {
  const x0 = getPoint(datum, seriesProps.x0Accessor, xAxis.type);
  if (x0 === void 0) {
    return void 0;
  }
  const x1 = seriesProps.x1Accessor && getPoint(datum, seriesProps.x1Accessor, xAxis.type);
  if (x1 === void 0 && !isCategorical(xAxis.type)) {
    return void 0;
  }
  const y0 = getPoint(datum, seriesProps.y0Accessor, yAxis.type);
  if (y0 === void 0) {
    return void 0;
  }
  const y1 = seriesProps.y1Accessor && getPoint(datum, seriesProps.y1Accessor, yAxis.type);
  if (y1 === void 0 && !isCategorical(yAxis.type)) {
    return void 0;
  }
  const value = getDatumValue(
    datum,
    seriesProps.valueAccessor
  );
  return {
    ...datum,
    value,
    x0,
    x1,
    y0,
    y1
  };
}
function getLineDatapoint(datum, seriesProps, xAxis, yAxis) {
  return getBaseDatapoint(datum, seriesProps, xAxis, yAxis);
}
const getAxisType = (axisName, xAxis, yAxis) => {
  const axisType = {
    x: xAxis,
    y: yAxis
  };
  return axisType[axisName].type;
};
const getChartOrientation = ({
  x1Accessor,
  x0Accessor,
  y1Accessor,
  y0Accessor
}) => {
  const isVertical = !!(x0Accessor && y1Accessor);
  const isHorizontal = !!(x1Accessor && y0Accessor);
  const isFullSquare = isVertical && isHorizontal;
  if (isFullSquare) {
    return "both";
  }
  if (isVertical) {
    return "vertical";
  }
  if (isHorizontal) {
    return "horizontal";
  } else {
    return void 0;
  }
};
const getChartAreaVariant = (chartOrientation, seriesProps) => {
  if (chartOrientation === "both") {
    return "band";
  }
  if (chartOrientation === "horizontal") {
    return seriesProps.x0Accessor ? "band" : "area";
  }
  return seriesProps.y0Accessor ? "band" : "area";
};
export {
  getAxisType,
  getBaseDatapoint,
  getChartAreaVariant,
  getChartOrientation,
  getDatumValue,
  getDotDatapoint,
  getLineDatapoint,
  getPoint,
  getRectDatapoint
};
