// @dynatrace/strato-components/filters/timeframe-selector/components/overlay/TimeframeSelectorOverlayFooter.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TimeframeSelectorOverlayFooter_exports = {};
__export(TimeframeSelectorOverlayFooter_exports, {
  TimeframeSelectorOverlayFooter: () => TimeframeSelectorOverlayFooter
});
module.exports = __toCommonJS(TimeframeSelectorOverlayFooter_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_tz = require("@date-fns/tz");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_spacings = __toESM(require("@dynatrace/strato-design-tokens/spacings"));
var import_user_preferences = require("@dynatrace-sdk/user-preferences");
var styles = __toESM(require("./TimeframeSelectorOverlay.sty.js"));
var import_Button = require("../../../../buttons/button/Button.js");
var import_external_links = require("../../../../core/external-links.js");
var import_useTimeframeStore = require("../../../../core/time/timeframeStore/useTimeframeStore.js");
var import_helpers = require("../../../../forms/forms-core/time/helpers.js");
var import_Divider = require("../../../../layouts/divider/Divider.js");
var import_Flex = require("../../../../layouts/flex/Flex.js");
var import_safe_sprinkles = require("../../../../styles/safe-sprinkles.js");
var import_ExternalLink = require("../../../../typography/external-link/ExternalLink.js");
var import_Text = require("../../../../typography/text/Text.js");
var import_timeframeSelectorMessages = require("../../constants/timeframeSelectorMessages.js");
const TimeframeSelectorOverlayFooter = (props) => {
  const { clearable } = props;
  const intl = (0, import_react_intl.useIntl)();
  const setValue = (0, import_useTimeframeStore.useTimeframeSelectorStore)(
    (state) => state.setTimeframeSelectorTimeframe
  );
  const timeZoneWithOffset = (0, import_react.useMemo)(() => {
    const timeZone = (0, import_user_preferences.getTimezone)();
    const parts = new Intl.DateTimeFormat((0, import_helpers.getSafeRegionalFormat)(), {
      timeZone,
      timeZoneName: "shortOffset"
    }).formatToParts(import_tz.TZDateMini.tz(timeZone));
    const partsShortName = new Intl.DateTimeFormat((0, import_helpers.getSafeRegionalFormat)(), {
      timeZone,
      timeZoneName: "short"
    }).formatToParts(import_tz.TZDateMini.tz(timeZone));
    const timeZonePart = parts?.find((part) => part.type === "timeZoneName")?.value.replace("GMT", "UTC");
    const timeZoneInfo = timeZonePart === "UTC" ? "UTC+0" : timeZonePart;
    const timezoneDetails = timeZoneInfo ? ` (${timeZoneInfo})` : "";
    const timeZoneShortName = partsShortName?.find(
      (part) => part.type === "timeZoneName"
    )?.value;
    return `${timeZoneShortName}${timezoneDetails}`;
  }, []);
  function handleClearSelection() {
    setValue(null);
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Divider.Divider, {}),
    /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      import_Flex.Flex,
      {
        justifyContent: "space-between",
        alignItems: "center",
        flexWrap: "wrap-reverse",
        flex: 1,
        gap: 4,
        className: styles.footer,
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
            import_Flex.Flex,
            {
              justifyContent: clearable ? "flex-start" : "space-between",
              alignItems: "center",
              flexShrink: 1,
              flexGrow: 1,
              className: (0, import_clsx.default)((0, import_safe_sprinkles.globalSprinkles)({ paddingLeft: 6 }), {
                [(0, import_safe_sprinkles.globalSprinkles)({ paddingTop: 4, paddingBottom: 4 })]: !clearable
              }),
              children: [
                /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { className: styles.hint, children: timeZoneWithOffset }),
                clearable && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Divider.Divider, { orientation: "vertical" }),
                /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  import_Text.Text,
                  {
                    as: import_ExternalLink.ExternalLink,
                    href: import_external_links.TIMEFRAME_SELECTOR_DOCUMENTATION_URL,
                    textStyle: "small",
                    style: { paddingRight: import_spacings.default.Size6 },
                    "data-dt-name": "LearnMore",
                    children: intl.formatMessage(import_timeframeSelectorMessages.TIMEFRAME_SELECTOR_MESSAGES.docsLinkText)
                  }
                )
              ]
            }
          ),
          clearable && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_Button.Button,
            {
              type: "reset",
              onPointerDown: handleClearSelection,
              size: "condensed",
              children: intl.formatMessage(
                import_timeframeSelectorMessages.TIMEFRAME_SELECTOR_MESSAGES.resetTimeframeButton
              )
            }
          )
        ]
      }
    )
  ] });
};
TimeframeSelectorOverlayFooter.displayName = "TimeframeSelectorOverlayFooter";
