// @dynatrace/strato-components/filters/timeframe-selector/components/overlay/TimeframeSelectorOverlay.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TimeframeSelectorOverlay_exports = {};
__export(TimeframeSelectorOverlay_exports, {
  TimeframeSelectorOverlay: () => TimeframeSelectorOverlay
});
module.exports = __toCommonJS(TimeframeSelectorOverlay_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_tz = require("@date-fns/tz");
var import_react = require("react");
var import_react_intl = require("react-intl");
var styles = __toESM(require("./TimeframeSelectorOverlay.sty.js"));
var import_TimeframeSelectorOverlayFooter = require("./TimeframeSelectorOverlayFooter.js");
var import_FocusScope = require("../../../../core/components/focus-scope/FocusScope.js");
var import_OverlayContainer = require("../../../../core/components/OverlayContainer.js");
var import_useOverlayStacking = require("../../../../core/hooks/useOverlayStacking.js");
var import_useRovingFocus = require("../../../../core/hooks/useRovingFocus.js");
var import_OverlayStackingProvider = require("../../../../core/providers/OverlayStackingProvider.js");
var import_useTimeframeStore = require("../../../../core/time/timeframeStore/useTimeframeStore.js");
var import_useDateTimeFormat = require("../../../../forms/date-time-picker/hooks/useDateTimeFormat.js");
var import_Flex = require("../../../../layouts/flex/Flex.js");
var import_useOverlay = require("../../../../overlays/hooks/useOverlay.js");
var import_timeframeSelectorMessages = require("../../constants/timeframeSelectorMessages.js");
var import_CustomTimeframeSelector = require("../custom-timeframe-selector/CustomTimeframeSelector.js");
var import_Presets = require("../presets/Presets.js");
var import_RecentTimeframes = require("../presets/RecentTimeframes.js");
function getMaxWidth(toParts) {
  const placeholderFromOrToParts = toParts(new import_tz.TZDateMini()).map(
    (part) => {
      if (part.type === "dayPeriod") {
        return "AM/PM";
      }
      return part.value;
    }
  );
  const placeholderFromOrTo = placeholderFromOrToParts.join("");
  const placeholdersNumberOfChars = 2 * placeholderFromOrTo.length;
  const surroundingSpace = 22;
  return `${placeholdersNumberOfChars * 1.05 + // try to compensate for the ch unit being calculated for "0" instead of a placeholder char, which needs slightly more width
  surroundingSpace}ch`;
}
const TimeframeSelectorOverlay = (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    overlayProps,
    /* eslint-disable @typescript-eslint/no-deprecated */
    clearable,
    children,
    overlayClose,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow
  } = props;
  const intl = (0, import_react_intl.useIntl)();
  const {
    trapFocus,
    style: overlayPropsStyle,
    ...remainingOverlayProps
  } = overlayProps;
  const overlayContainerProps = (0, import_useOverlayStacking.useOverlayStacking)(void 0, "overlay");
  const precision = (0, import_useTimeframeStore.useTimeframeSelectorStore)((state) => state.precision);
  const { toParts } = (0, import_useDateTimeFormat.useDateTimeFormat)("datetime", precision);
  const { containerRef } = (0, import_useRovingFocus.useRovingFocus)({
    initialIndex: 0,
    direction: "vertical",
    focusableChildrenSelector: "[role='option'][aria-disabled='false']"
  });
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_OverlayContainer.OverlayContainer,
    {
      ...overlayContainerProps,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayStackingProvider.OverlayStackingProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_FocusScope.FocusScope, { autoFocus: true, children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
          import_Flex.Flex,
          {
            ref: forwardedRef,
            gap: 4,
            flexDirection: "column",
            padding: 4,
            className: styles.wrapper,
            style: {
              ...overlayPropsStyle,
              width: getMaxWidth(toParts),
              maxHeight: `min(${overlayPropsStyle?.maxHeight}, 600px)`
            },
            "aria-label": intl.formatMessage(
              import_timeframeSelectorMessages.TIMEFRAME_SELECTOR_MESSAGES.overlayContainer
            ),
            ...remainingOverlayProps,
            children: [
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_CustomTimeframeSelector.CustomTimeframeSelector, {}),
              /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
                import_Flex.Flex,
                {
                  ref: containerRef,
                  gap: 4,
                  paddingLeft: 0,
                  paddingRight: 0,
                  flexDirection: "column",
                  className: styles.scrollContainer,
                  children: [
                    children ?? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Presets.Presets, {}),
                    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_RecentTimeframes.RecentTimeframes, {})
                  ]
                }
              ),
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TimeframeSelectorOverlayFooter.TimeframeSelectorOverlayFooter, { clearable })
            ]
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_useOverlay.DismissButton, { onDismiss: overlayClose })
      ] }) })
    }
  );
});
TimeframeSelectorOverlay.displayName = "TimeframeSelectorOverlay";
