// @dynatrace/strato-components/esm/charts/xy-chart/components/tooltip/XYChartTooltipTemplate/XYChartTooltipTemplate.js
import { jsx } from "react/jsx-runtime";
import { forwardRef } from "react";
import { body } from "./XYChartTooltipTemplate.sty.js";
import { Flex } from "../../../../../layouts/flex/Flex.js";
import { useChartActionsMenu } from "../../../../core/components/series-action/hooks/useChartActionsMenu.js";
import {
  ChartTooltip
} from "../../../../core/components/tooltip/ChartTooltip/ChartTooltip.js";
const XYChartTooltipTemplate = forwardRef(({ position, pinned, children }, ref) => {
  const {
    selectedItem: selectedItemId,
    actionsMenuRef,
    updateSelectedItem,
    onLeave
  } = useChartActionsMenu();
  if (!position) {
    return null;
  }
  return /* @__PURE__ */ jsx(
    ChartTooltip,
    {
      ref,
      position,
      pinned,
      className: body,
      updatePositionWhenPinned: true,
      onMouseLeave: onLeave,
      actionSelectionValue: {
        selectedItemId,
        actionsMenuRef,
        updateSelectedItem
      },
      children: /* @__PURE__ */ jsx(Flex, { flexDirection: "column", gap: 4, children })
    }
  );
});
XYChartTooltipTemplate.displayName = "XYChartTooltipTemplate";
export {
  XYChartTooltipTemplate
};
