// @dynatrace/strato-components/forms/_base-input/baseInputMessages.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var baseInputMessages_exports = {};
__export(baseInputMessages_exports, {
  getPlaceholderMessage: () => getPlaceholderMessage
});
module.exports = __toCommonJS(baseInputMessages_exports);
var import_react_intl = require("react-intl");
const PLACEHOLDER_MESSAGES = (0, import_react_intl.defineMessages)({
  text: {
    defaultMessage: "Add text",
    description: "Default placeholder for text input.",
    id: "7tnC6f9ag1/zG/vh"
  },
  password: {
    defaultMessage: "Add password",
    description: "Default placeholder for password input.",
    id: "B3Rj3sRBMMfDBuzK"
  },
  email: {
    defaultMessage: "Add email",
    description: "Default placeholder for email input.",
    id: "xcSpnebtYoA/jEJY"
  },
  url: {
    defaultMessage: "Add URL",
    description: "Default placeholder for URL input.",
    id: "IK8qCfDGg9HgIHQC"
  },
  tel: {
    defaultMessage: "Add phone number",
    description: "Default placeholder for phone number input.",
    id: "xjwKUTJHlmtA/V/n"
  },
  search: {
    defaultMessage: "Search",
    description: "Default placeholder for search input.",
    id: "Fr3cNB5JDQGaGwae"
  }
});
function isPlaceholderType(type) {
  return type !== void 0 && type in PLACEHOLDER_MESSAGES;
}
function getPlaceholderMessage(type) {
  if (isPlaceholderType(type)) {
    return PLACEHOLDER_MESSAGES[type];
  }
  return PLACEHOLDER_MESSAGES.text;
}
