// @dynatrace/strato-components/forms/_base-input/BaseInputButton.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var BaseInputButton_exports = {};
__export(BaseInputButton_exports, {
  BaseInputButton: () => BaseInputButton
});
module.exports = __toCommonJS(BaseInputButton_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_useBaseInputContext = require("./hooks/useBaseInputContext.js");
var import_Button = require("../../buttons/button/Button.js");
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useId = require("../../core/hooks/useId.js");
var import_useSafeForwardProps = require("../../core/hooks/useSafeForwardProps.js");
var import_merge_props = require("../../core/utils/merge-props.js");
var import_Tooltip = require("../../overlays/tooltip/Tooltip.js");
const BaseInputButton = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    children,
    tooltipText,
    tooltipBehavioralTrackingProps: tooltipTrackingProps,
    disabled: propDisabled,
    id,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "aria-disabled": ariaDisabledProp,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restPropsWithoutTracking] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const {
    hasError = false,
    disabled,
    variant,
    ariaDisabled: contextAriaDisabled
  } = (0, import_useBaseInputContext.useBaseInputContext)();
  const ariaDisabled = ariaDisabledProp || contextAriaDisabled;
  const tooltipGeneratedId = (0, import_useId.useId)();
  const tooltipId = tooltipText ? id ?? tooltipGeneratedId : void 0;
  const { "aria-labelledby": ariaLabelledByProp, ...remainingAriaProps } = (0, import_useAriaLabelingProps.useAriaLabelingProps)(restPropsWithoutTracking);
  const forwardProps = (0, import_useSafeForwardProps.useSafeForwardProps)(restPropsWithoutTracking, {
    ariaLabelledByProp,
    ...remainingAriaProps
  });
  const ariaLabelledBy = (0, import_react.useMemo)(
    () => tooltipId || ariaLabelledByProp ? `${tooltipId ? tooltipId + " " : ""}${ariaLabelledByProp ?? ""}` : void 0,
    [ariaLabelledByProp, tooltipId]
  );
  const button = (0, import_react.useMemo)(
    () => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Button.Button,
      {
        ref: forwardedRef,
        ...forwardProps,
        variant: variant === "minimal" ? "default" : "emphasized",
        disabled: propDisabled ?? disabled,
        "aria-disabled": ariaDisabled,
        color: hasError && !disabled ? "critical" : "neutral",
        "aria-labelledby": ariaLabelledBy,
        className: consumerClassName,
        style: consumerStyle,
        "data-testid": dataTestId,
        ...(0, import_merge_props.mergeProps)(remainingAriaProps, behavioralTrackingProps),
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children })
      }
    ),
    [
      ariaLabelledBy,
      behavioralTrackingProps,
      children,
      consumerClassName,
      consumerStyle,
      dataTestId,
      disabled,
      forwardProps,
      forwardedRef,
      hasError,
      propDisabled,
      remainingAriaProps,
      variant,
      ariaDisabled
    ]
  );
  return tooltipText ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Tooltip.Tooltip,
    {
      id: tooltipId,
      text: tooltipText,
      placement: "top-end",
      ...tooltipTrackingProps,
      children: button
    }
  ) : button;
});
BaseInputButton.displayName = "BaseInput.Button";
