// @dynatrace/strato-components/forms/_base-input/BaseInput.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var BaseInput_exports = {};
__export(BaseInput_exports, {
  BaseInput: () => BaseInput
});
module.exports = __toCommonJS(BaseInput_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_intl = require("react-intl");
var styles = __toESM(require("./BaseInput.sty.js"));
var import_BaseInputButton = require("./BaseInputButton.js");
var import_BaseInputContent = require("./BaseInputContent.js");
var import_BaseInputContext = require("./contexts/BaseInputContext.js");
var import_useFormControlState = require("../../core/hooks/useFormControlState.js");
var import_useSplitProps = require("../../core/hooks/useSplitProps.js");
var import_useFocusRing = require("../../core/styles/useFocusRing.js");
var import_is_slot = require("../../core/utils/is-slot.js");
var import_merge_props = require("../../core/utils/merge-props.js");
var import_Divider = require("../../layouts/divider/Divider.js");
var import_Flex = require("../../layouts/flex/Flex.js");
var import_InputGroup_css = require("../../layouts/input-group/InputGroup.sty.js");
var import_InputGroup = require("../../layouts/input-group/InputGroup.js");
var import_FormFieldMessageBaseItem = require("../forms-core/components/form-field-messages/FormFieldMessageBaseItem.js");
var import_useFormFieldMessagesContext = require("../forms-core/hooks/useFormFieldMessagesContext.js");
var import_useFormFieldValidation = require("../forms-core/hooks/useFormFieldValidation.js");
var import_useSyncedFormControlProps = require("../forms-core/hooks/useSyncedFormControlProps.js");
var import_formfield_messages_state_reducer = require("../forms-core/state/formfield-messages-state-reducer.js");
var import_useAutofillEvent = require("../hooks/useAutofillEvent.js");
var import_useInputHint = require("../hooks/useInputHint.js");
var import_baseInputMessages = require("./baseInputMessages.js");
var import_ClearButton = require("./ClearButton.js");
function getSlots(children) {
  const validSlots = import_react.Children.toArray(children).filter(
    import_react.isValidElement
  );
  let prefix = { buttons: [], rest: [], props: {} };
  let suffix = { buttons: [], rest: [], props: {} };
  let reservedSuffix = { buttons: [], rest: [], props: {} };
  for (const slot of validSlots) {
    const buttons = [];
    const rest = [];
    const slotChildren = import_react.Children.toArray(slot.props.children);
    for (const child of slotChildren) {
      if ((0, import_is_slot.isSlot)(child, import_BaseInputButton.BaseInputButton)) {
        buttons.push(child);
      } else {
        rest.push(child);
      }
    }
    const { children: _children, ...slotProps } = slot.props;
    if (slot.type === import_BaseInputContent.BaseInputPrefix) {
      prefix = { buttons, rest, props: slotProps };
    }
    if (slot.type === import_BaseInputContent.BaseInputSuffix) {
      suffix = { buttons, rest, props: slotProps };
    }
    if (slot.type === import_BaseInputContent.BaseInputReservedSuffix) {
      reservedSuffix = { buttons, rest, props: slotProps };
    }
  }
  return { prefix, suffix, reservedSuffix };
}
const BaseInputRoot = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const intl = (0, import_react_intl.useIntl)();
  const defaultPlaceholder = intl.formatMessage(
    (0, import_baseInputMessages.getPlaceholderMessage)(props.type)
  );
  const {
    children,
    id: propId,
    disabled: propDisabled,
    required: propRequired,
    "aria-disabled": ariaDisabledProp,
    variant = "default",
    controlState,
    readOnly: propReadOnly,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "aria-label": ariaLabelProp,
    "aria-labelledby": ariaLabelledByProp,
    "aria-describedby": ariaDescribedByProp,
    "aria-details": ariaDetailsProp,
    className: consumerClassName,
    style: consumerStyle,
    // eslint-disable-next-line @typescript-eslint/no-deprecated
    onClear,
    type,
    placeholder = defaultPlaceholder,
    value,
    pattern,
    max,
    min,
    step,
    maxLength,
    minLength,
    ...remainingProps
  } = props;
  const inputValue = value ?? "";
  const { disabled, required, id, ariaDisabled, readOnly } = (0, import_useSyncedFormControlProps.useSyncedFormControlProps)({
    propDisabled,
    propRequired,
    propId,
    propAriaDisabled: ariaDisabledProp,
    propReadOnly
  });
  const disabledOrAriaDisabled = disabled || ariaDisabled || false;
  const [hasFormControlError, hint] = (0, import_useFormControlState.useFormControlState)(controlState);
  const { inputProps, hintProps } = (0, import_useInputHint.useInputHint)({
    hint,
    hasError: hasFormControlError
  });
  const { state, dispatch } = (0, import_useFormFieldMessagesContext.useFormFieldMessagesContext)();
  const hasError = hasFormControlError || state.showError;
  const { isAutoFilled, onAnimationStart } = (0, import_useAutofillEvent.useAutofillEvent)();
  const inputRef = (0, import_react.useRef)(null);
  const wrapperRef = (0, import_react.useRef)(null);
  const { focusProps, focusClassName } = (0, import_useFocusRing.useFocusRing)({
    color: hasError && !(disabled || ariaDisabled) ? "critical" : "neutral",
    ignoreModality: true,
    disabled
  });
  const { onChange, onBlur, onInvalid, ariaProps } = (0, import_useFormFieldValidation.useFormFieldValidation)(
    inputRef,
    {
      required,
      pattern,
      max,
      min,
      step,
      maxLength,
      minLength,
      type,
      value: inputValue
    },
    {
      "aria-label": ariaLabelProp,
      "aria-labelledby": ariaLabelledByProp,
      "aria-describedby": ariaDescribedByProp,
      "aria-details": ariaDetailsProp
    }
  );
  const eventListener = { onChange, onBlur, onInvalid };
  (0, import_react.useImperativeHandle)(
    forwardedRef,
    () => ({
      element: wrapperRef.current,
      inputRef: inputRef.current,
      focus() {
        inputRef.current?.focus();
      },
      validate() {
        inputRef.current?.checkValidity();
        dispatch?.({
          type: import_formfield_messages_state_reducer.FormFieldMessagesAction.SETISTOUCHED,
          payload: { isTouched: true }
        });
      }
    }),
    [dispatch]
  );
  const context = (0, import_react.useMemo)(
    () => ({ variant, hasError, disabled, ariaDisabled }),
    [variant, hasError, disabled, ariaDisabled]
  );
  const slots = (0, import_react.useMemo)(() => getSlots(children), [children]);
  const hasInnerPrefix = slots.prefix.rest.length > 0;
  const hasInnerSuffix = slots.suffix.rest.length > 0;
  const hasInnerReservedSuffix = slots.reservedSuffix.rest.length > 0;
  const hasOnClear = onClear !== void 0;
  const showClearButton = hasOnClear && inputValue !== "" && !hasInnerReservedSuffix;
  const {
    matchedProps: dataAttributesProps,
    remainingProps: remainingInputProps
  } = (0, import_useSplitProps.useSplitProps)(remainingProps, [(key) => key.startsWith("data-")]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_Flex.Flex,
    {
      gap: 4,
      flexDirection: "column",
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ref: wrapperRef,
      className: consumerClassName,
      style: consumerStyle,
      ...dataAttributesProps,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_InputGroup.InputGroup,
          {
            style: {
              gridAutoColumns: `${"auto ".repeat(
                slots.prefix.buttons.length
              )}1fr${" auto".repeat(slots.suffix.buttons.length)}`
            },
            className: (0, import_clsx.default)(styles.wrapper, styles.variant[variant], {
              [styles.hasError]: hasError,
              [styles.disabled]: disabledOrAriaDisabled,
              [styles.readOnly]: readOnly
            }),
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_BaseInputContext.BaseInputContext.Provider, { value: context, children: [
              slots.prefix.buttons,
              /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
                "div",
                {
                  className: (0, import_clsx.default)(
                    styles.container,
                    import_InputGroup_css.inputGroupClassName,
                    focusClassName,
                    {
                      [styles.isAutoFilled]: isAutoFilled,
                      [`${styles.containerPadding.none}`]: hasInnerReservedSuffix && hasInnerPrefix,
                      [`${styles.containerPadding.rightNone}`]: hasInnerReservedSuffix && !hasInnerPrefix,
                      [`${styles.containerPadding.leftUnset}`]: hasInnerPrefix,
                      [`${styles.containerPadding.rightUnset}`]: hasInnerSuffix
                    }
                  ),
                  onClick: () => inputRef.current?.focus(),
                  children: [
                    hasInnerPrefix && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_BaseInputContent.BaseInputPrefix, { ...slots.prefix.props, children: slots.prefix.rest }),
                    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                      "input",
                      {
                        id,
                        ref: inputRef,
                        required,
                        disabled: !!disabled,
                        "aria-disabled": ariaDisabled,
                        onAnimationStart,
                        readOnly,
                        className: styles.input,
                        placeholder,
                        type,
                        value: inputValue,
                        pattern,
                        max,
                        min,
                        step,
                        maxLength,
                        minLength,
                        ...(0, import_merge_props.mergeProps)(
                          remainingInputProps,
                          focusProps,
                          controlState ? { "aria-label": ariaLabelProp, ...inputProps } : ariaProps,
                          !controlState ? eventListener : {}
                          // escape hatch to not interfere with the old implementation
                        )
                      }
                    ),
                    hasInnerSuffix && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_BaseInputContent.BaseInputSuffix, { ...slots.suffix.props, children: slots.suffix.rest }),
                    !readOnly && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, { children: hasInnerReservedSuffix ? /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
                      hasInnerSuffix && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                        import_Divider.Divider,
                        {
                          orientation: "vertical",
                          className: styles.divider
                        }
                      ),
                      slots.reservedSuffix.rest
                    ] }) : /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
                      hasInnerSuffix && showClearButton && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                        import_Divider.Divider,
                        {
                          orientation: "vertical",
                          className: styles.divider
                        }
                      ),
                      showClearButton && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ClearButton.ClearButton, { onClear })
                    ] }) })
                  ]
                }
              ),
              slots.suffix.buttons
            ] })
          }
        ),
        controlState && hint && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_FormFieldMessageBaseItem.InternalBaseItem,
          {
            variant: hasError ? "error" : "hint",
            disabled: disabledOrAriaDisabled,
            ...hintProps,
            children: hint
          }
        )
      ]
    }
  );
});
const BaseInput = Object.assign(BaseInputRoot, {
  Prefix: import_BaseInputContent.BaseInputPrefix,
  Suffix: import_BaseInputContent.BaseInputSuffix,
  ReservedSuffix: import_BaseInputContent.BaseInputReservedSuffix,
  Button: import_BaseInputButton.BaseInputButton
});
BaseInput.displayName = "BaseInput";
