// @dynatrace/strato-components/forms/_base-input/BaseInput.sty.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var BaseInput_css_exports = {};
__export(BaseInput_css_exports, {
  container: () => container,
  containerPadding: () => containerPadding,
  disabled: () => disabled,
  divider: () => divider,
  hasError: () => hasError,
  input: () => input,
  isAutoFilled: () => isAutoFilled,
  readOnly: () => readOnly,
  reservedSuffixButton: () => reservedSuffixButton,
  variant: () => variant,
  wrapper: () => wrapper
});
module.exports = __toCommonJS(BaseInput_css_exports);
var import_AppRoot_css_ts_vanilla = require("../../core/components/app-root/AppRoot.css");
var import_field_css_ts_vanilla = require("../../styles/field.css");
var import_textStyle_css_ts_vanilla = require("../../styles/textStyle.css");
var import_useAutofillEvent_css_ts_vanilla = require("../hooks/useAutofillEvent.css");
var import_BaseInput_css_ts_vanilla = require("./BaseInput.css");
var container = "_1a8lqy98-3-6-0";
var containerPadding = { none: "_1a8lqy99-3-6-0", rightNone: "_1a8lqy9a-3-6-0", rightUnset: "_1a8lqy9b-3-6-0", leftUnset: "_1a8lqy9c-3-6-0" };
var disabled = "_1a8lqy92-3-6-0";
var divider = "_1a8lqy9f-3-6-0";
var hasError = "_1a8lqy91-3-6-0";
var input = "_1a8lqy9d-3-6-0 _yfwka33-3-6-0 _da9a8v0-3-6-0 _da9a8v2-3-6-0 _da9a8v3-3-6-0 _da9a8vb-3-6-0";
var isAutoFilled = "_1a8lqy9g-3-6-0";
var readOnly = "_1a8lqy93-3-6-0";
var reservedSuffixButton = "_1a8lqy9e-3-6-0";
var variant = { "default": "_1a8lqy94-3-6-0", minimal: "_1a8lqy95-3-6-0" };
var wrapper = "_1a8lqy90-3-6-0";
