// @dynatrace/strato-components/forms/_base-input/ClearButton.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ClearButton_exports = {};
__export(ClearButton_exports, {
  ClearButton: () => ClearButton
});
module.exports = __toCommonJS(ClearButton_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_strato_icons = require("@dynatrace/strato-icons");
var baseInputStyles = __toESM(require("./BaseInput.sty.js"));
var import_BaseInputContext = require("./contexts/BaseInputContext.js");
var import_Button = require("../../buttons/button/Button.js");
const ClearButton = ({ onClear }) => {
  const { hasError, disabled, ariaDisabled } = (0, import_react.useContext)(import_BaseInputContext.BaseInputContext);
  const disabledOrAriaDisabled = disabled || ariaDisabled || false;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Button.Button,
    {
      onClick: onClear,
      className: baseInputStyles.reservedSuffixButton,
      color: hasError ? "critical" : "neutral",
      disabled: disabledOrAriaDisabled,
      "data-dt-name": "ClearAll",
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.XmarkIcon, {})
    }
  );
};
