// @dynatrace/strato-components/esm/forms/forms-core/state/formfield-messages-state-reducer.js
var FormFieldMessagesAction = /* @__PURE__ */ ((FormFieldMessagesAction2) => {
  FormFieldMessagesAction2["ADDMESSAGE"] = "addmessage";
  FormFieldMessagesAction2["REMOVEMESSAGE"] = "removemessage";
  FormFieldMessagesAction2["DISABLEMESSAGE"] = "disablemessage";
  FormFieldMessagesAction2["SETISTOUCHED"] = "setistouched";
  FormFieldMessagesAction2["ADDFORMCONTROLID"] = "addformcontrolid";
  FormFieldMessagesAction2["REMOVEFORMCONTROLID"] = "removeformcontrolid";
  return FormFieldMessagesAction2;
})(FormFieldMessagesAction || {});
function isFormControlPayload(payload, messageType) {
  return messageType === "formcontrol";
}
function isHTMLFormControlElement(element) {
  return !!(element && "setCustomValidity" in element);
}
function setHTMLElementValidity(id, isValid) {
  const element = document.getElementById(id);
  if (isHTMLFormControlElement(element)) {
    element.setCustomValidity(isValid ? "" : " ");
  }
}
function updateValidity(state, explicitControlId) {
  const { customMessages, formControlMessages, isTouched, isDisabled } = state;
  const hasCustomError = customMessages.some(
    (message) => message.variant === "error"
  );
  const hasFormControlError = formControlMessages.some(
    (message) => message.variant === "error"
  );
  const newIsValid = !hasCustomError && !hasFormControlError;
  const newShowError = !isDisabled && (state.showError || (isTouched ?? false)) && !newIsValid;
  const hasStratoCustomError = formControlMessages.some(
    (message) => message.variant === "error" && message.validityStateKey === "customError"
  );
  const nativeValidityIsValid = !hasCustomError && !hasStratoCustomError;
  if (explicitControlId) {
    setHTMLElementValidity(explicitControlId, nativeValidityIsValid);
    return {
      ...state,
      isValid: newIsValid,
      showError: newShowError
    };
  }
  state.controlIds?.forEach(
    (id) => setHTMLElementValidity(id, nativeValidityIsValid)
  );
  return {
    ...state,
    isValid: newIsValid,
    showError: newShowError
  };
}
function formFieldMessagesReducer(state, action) {
  const newState = { ...state };
  if (action.type === "addmessage" /* ADDMESSAGE */) {
    const { payload } = action;
    const { messageType, ...remainingPayload } = payload;
    if (isFormControlPayload(remainingPayload, messageType)) {
      if (remainingPayload.validityStateKey === "customError" && remainingPayload.variant === "error" && "id" in remainingPayload) {
        const { id: internalId, ...internalRest } = remainingPayload;
        const customErrors = state.formControlMessages.filter(
          (m) => m.validityStateKey === "customError"
        );
        if (customErrors.length === 1 && customErrors[0].id === internalId) {
          if ((payload.showError ?? state.showError) === state.showError) {
            return state;
          }
        } else {
          newState.formControlMessages = [
            ...state.formControlMessages.filter(
              (m) => m.validityStateKey !== "customError"
            ),
            { id: internalId, ...internalRest }
          ];
        }
      } else {
        const isNewBrowserError = remainingPayload.validityStateKey !== void 0 && !state.formControlMessages.some(
          (message) => message.validityStateKey === remainingPayload.validityStateKey
        );
        if (!isNewBrowserError) {
          return state;
        }
        newState.formControlMessages = [
          ...newState.formControlMessages,
          { id: `strato-${crypto.randomUUID()}`, ...remainingPayload }
        ];
      }
    } else if ("id" in remainingPayload && !state.formControlMessages.some(
      (message) => message.id === remainingPayload.id
    ) && !state.customMessages.some(
      (message) => message.id === remainingPayload.id
    )) {
      newState.customMessages.push(remainingPayload);
    }
    if (remainingPayload.variant === "error") {
      newState.showError = payload.showError ?? state.showError;
    }
    return updateValidity(newState, payload.controlId);
  }
  if (action.type === "removemessage" /* REMOVEMESSAGE */) {
    const { payload } = action;
    let didUpdate = false;
    if (payload.messageType === "formcontrol") {
      newState.formControlMessages = newState.formControlMessages.filter(
        (message) => {
          if (message.id === payload.id) {
            didUpdate = true;
          }
          return message.id !== payload.id;
        }
      );
    } else {
      newState.customMessages = newState.customMessages.filter((message) => {
        if (message.id === payload.id) {
          didUpdate = true;
        }
        return message.id !== payload.id;
      });
    }
    if (!didUpdate) {
      return state;
    }
    return updateValidity(newState, payload.controlId);
  }
  if (action.type === "disablemessage" /* DISABLEMESSAGE */) {
    const { payload } = action;
    newState.isDisabled = payload.isDisabled;
  }
  if (action.type === "setistouched" /* SETISTOUCHED */) {
    const { payload } = action;
    newState.isTouched = payload.isTouched;
  }
  if (action.type === "addformcontrolid" /* ADDFORMCONTROLID */) {
    const {
      payload: { controlId }
    } = action;
    newState.controlIds ??= [];
    if (!newState.controlIds.includes(controlId)) {
      newState.controlIds.push(controlId);
    }
  }
  if (action.type === "removeformcontrolid" /* REMOVEFORMCONTROLID */) {
    const {
      payload: { controlId }
    } = action;
    newState.controlIds = state.controlIds?.filter((id) => id !== controlId);
  }
  return updateValidity(newState);
}
export {
  FormFieldMessagesAction,
  formFieldMessagesReducer
};
