// @dynatrace/strato-components/esm/charts/pie/DonutChart.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import { isEmpty } from "lodash-es";
import {
  forwardRef,
  useImperativeHandle,
  useRef
} from "react";
import { PieChartInternal } from "./components/PieChartInternal.js";
import {
  CHART_VARIANT,
  DONUT_CHART_ARIA_LABEL,
  DONUT_CHART_DISPLAY_NAME
} from "./constants.js";
import { InnerElementContext } from "./context/InnerElementContext.js";
import { usePieChartConfig } from "./hooks/usePieChartConfig.js";
import {
  ChartProviders
} from "./providers/ChartProviders.js";
import { Toolbar } from "./shared/slots/Toolbar.js";
import { DonutChartDownloadCSVSlot } from "./slots/DonutChartDownloadCSV.js";
import { Grouping as DonutChartGrouping } from "./slots/DonutChartGrouping.js";
import { DonutChartInner } from "./slots/DonutChartInner.js";
import { Labels as DonutChartLabels } from "./slots/DonutChartLabels.js";
import { Legend as DonutChartLegend } from "./slots/DonutChartLegend.js";
import { Tooltip as DonutChartTooltipSlot } from "./slots/DonutChartTooltip.js";
import { buildAndDownloadCsv } from "./utils/build-and-download-csv.js";
import { useMergeRefs } from "../../core/hooks/useMergeRefs.js";
import { Intent } from "../../core/slots/Intent/Intent.js";
import { defaultChartSize } from "../../core/styles/defaultChartValues.sty.js";
import { getChartStatesTemplates } from "../core/components/chart-states/utils/get-chart-states-template.js";
import { getToolbarConfigWithoutIntents } from "../core/components/toolbar/utils/get-toolbar-config-without-intents.js";
import { EmptyStateSlot } from "../core/slots/chart-states/EmptyStateSlot.js";
import { ErrorStateSlot } from "../core/slots/chart-states/ErrorStateSlot.js";
import { ColorRule } from "../core/slots/ColorRule.js";
import { ChartType } from "../core/types/chart-types.js";
const COMPONENT_NAME = "DonutChart";
const sanitizeToolbarIntents = (config) => {
  const toolbar = config.toolbar ? getToolbarConfigWithoutIntents(config.toolbar) : void 0;
  return {
    ...config,
    toolbar
  };
};
const DonutChartRoot = forwardRef((props, forwardedRef) => {
  const {
    className,
    style,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "data-testid": dataTestId = "donut-chart-root"
  } = props;
  const chartConfig = usePieChartConfig(props);
  const { emptyState, errorState } = getChartStatesTemplates(props.children);
  const isDataAvailable = !isEmpty(props.data);
  const containerRef = useRef(null);
  const rootRef = useMergeRefs([
    forwardedRef,
    containerRef
  ]);
  useImperativeHandle(
    forwardedRef,
    () => ({
      element: containerRef.current,
      getConfig: () => JSON.stringify({
        type: ChartType.Donut,
        ...sanitizeToolbarIntents(chartConfig)
      }),
      downloadData: () => {
        buildAndDownloadCsv(props.data);
      }
    })
  );
  const providersProps = {
    chartData: props.data,
    chartConfig
  };
  return /* @__PURE__ */ jsx(
    InnerElementContext.Provider,
    {
      value: {
        innerElement: chartConfig.innerElement
      },
      children: /* @__PURE__ */ jsx(ChartProviders, { ...providersProps, children: /* @__PURE__ */ jsx(
        PieChartInternal,
        {
          ...props,
          ref: rootRef,
          variant: CHART_VARIANT.Donut,
          isEmpty: !isDataAvailable,
          emptyState,
          errorState,
          className: clsx(defaultChartSize, className),
          style,
          "data-dtrum-mask": dataDtrumMask,
          "data-dtrum-allow": dataDtrumAllow,
          "data-testid": dataTestId,
          "data-dt-component": COMPONENT_NAME,
          "aria-label": DONUT_CHART_ARIA_LABEL,
          role: "img"
        }
      ) })
    }
  );
});
DonutChartRoot.displayName = DONUT_CHART_DISPLAY_NAME;
const DonutChart = Object.assign(DonutChartRoot, {
  Legend: DonutChartLegend,
  Grouping: DonutChartGrouping,
  Labels: DonutChartLabels,
  Inner: DonutChartInner,
  Tooltip: DonutChartTooltipSlot,
  Intent,
  Toolbar,
  EmptyState: EmptyStateSlot,
  ErrorState: ErrorStateSlot,
  ColorRule,
  DownloadCSV: DonutChartDownloadCSVSlot
});
export {
  DonutChart
};
