// @dynatrace/strato-components/esm/charts/pie/PieChart.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import { isEmpty } from "lodash-es";
import {
  forwardRef,
  useImperativeHandle,
  useRef
} from "react";
import { PieChartInternal } from "./components/PieChartInternal.js";
import {
  CHART_VARIANT,
  PIE_CHART_ARIA_LABEL,
  PIE_CHART_DISPLAY_NAME
} from "./constants.js";
import { usePieChartConfig } from "./hooks/usePieChartConfig.js";
import {
  ChartProviders
} from "./providers/ChartProviders.js";
import { Grouping as PieChartGrouping } from "./slots/PieChartGrouping.js";
import { Labels as PieChartLabels } from "./slots/PieChartLabels.js";
import { Legend as PieChartLegend } from "./slots/PieChartLegend.js";
import { Tooltip as PieChartTooltipSlot } from "./slots/PieChartTooltip.js";
import { buildAndDownloadCsv } from "./utils/build-and-download-csv.js";
import { useMergeRefs } from "../../core/hooks/useMergeRefs.js";
import { Intent } from "../../core/slots/Intent/Intent.js";
import { defaultChartSize } from "../../core/styles/defaultChartValues.sty.js";
import { getChartStatesTemplates } from "../core/components/chart-states/utils/get-chart-states-template.js";
import { getToolbarConfigWithoutIntents } from "../core/components/toolbar/utils/get-toolbar-config-without-intents.js";
import { EmptyStateSlot } from "../core/slots/chart-states/EmptyStateSlot.js";
import { ErrorStateSlot } from "../core/slots/chart-states/ErrorStateSlot.js";
import { ColorRule } from "../core/slots/ColorRule.js";
import { ChartType } from "../core/types/chart-types.js";
import { Toolbar } from "./shared/slots/Toolbar.js";
import { PieChartDownloadCSVSlot } from "./slots/PieChartDownloadCSV.js";
const COMPONENT_NAME = "PieChart";
const sanitizeToolbarIntents = (config) => {
  const toolbar = config.toolbar ? getToolbarConfigWithoutIntents(config.toolbar) : void 0;
  return {
    ...config,
    toolbar
  };
};
const PieChartRoot = forwardRef(
  (props, forwardedRef) => {
    const {
      className,
      style,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "data-testid": dataTestId = "pie-chart-root"
    } = props;
    const chartConfig = usePieChartConfig(props);
    const { emptyState, errorState } = getChartStatesTemplates(props.children);
    const isDataAvailable = !isEmpty(props.data);
    const containerRef = useRef(null);
    const rootRef = useMergeRefs([
      forwardedRef,
      containerRef
    ]);
    useImperativeHandle(
      forwardedRef,
      () => ({
        element: containerRef.current,
        getConfig: () => JSON.stringify({
          type: ChartType.Pie,
          ...sanitizeToolbarIntents(chartConfig)
        }),
        downloadData: () => {
          buildAndDownloadCsv(props.data);
        }
      })
    );
    const providersProps = {
      chartData: props.data,
      chartConfig
    };
    return /* @__PURE__ */ jsx(ChartProviders, { ...providersProps, children: /* @__PURE__ */ jsx(
      PieChartInternal,
      {
        ...props,
        ref: rootRef,
        height: chartConfig.height,
        variant: CHART_VARIANT.Pie,
        isEmpty: !isDataAvailable,
        emptyState,
        errorState,
        className: clsx(defaultChartSize, className),
        style,
        "data-dt-component": COMPONENT_NAME,
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        "aria-label": PIE_CHART_ARIA_LABEL,
        role: "img"
      }
    ) });
  }
);
PieChartRoot.displayName = PIE_CHART_DISPLAY_NAME;
const PieChart = Object.assign(PieChartRoot, {
  Legend: PieChartLegend,
  Grouping: PieChartGrouping,
  Labels: PieChartLabels,
  Tooltip: PieChartTooltipSlot,
  Intent,
  EmptyState: EmptyStateSlot,
  ErrorState: ErrorStateSlot,
  ColorRule,
  Toolbar,
  DownloadCSV: PieChartDownloadCSVSlot
});
export {
  PieChart
};
