// @dynatrace/strato-components/charts/meter-bar/multi-meter-bar/components/MultiMeterBarChartSegment.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MultiMeterBarChartSegment_exports = {};
__export(MultiMeterBarChartSegment_exports, {
  MultiMeterBarChartSegment: () => MultiMeterBarChartSegment
});
module.exports = __toCommonJS(MultiMeterBarChartSegment_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_MeterBarSegment = require("../../shared/components/meter-bar-segment/MeterBarSegment.js");
var import_selectors = require("../hooks/store/selectors.js");
const MultiMeterBarChartSegment = (props) => {
  const {
    previousSegmentsWidth,
    segmentColor,
    segmentWidth,
    segmentId,
    segmentIndex
  } = props;
  const highlightedSegment = (0, import_selectors.useMultiMeterBarHighlightedSegment)();
  const highlightedSegmentId = highlightedSegment?.id;
  const isHighlighted = !highlightedSegment || highlightedSegmentId === segmentId;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("g", { "data-segment-index": segmentIndex, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_MeterBarSegment.MeterBarSegment,
    {
      color: segmentColor,
      width: segmentWidth,
      highlighted: isHighlighted,
      start: previousSegmentsWidth
    }
  ) });
};
MultiMeterBarChartSegment["displayName"] = "MultiMeterBarChartSegment";
