// @dynatrace/strato-components/charts/meter-bar/multi-meter-bar/components/MultiMeterBarChartSegments.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MultiMeterBarChartSegments_exports = {};
__export(MultiMeterBarChartSegments_exports, {
  MultiMeterBarChartSegments: () => MultiMeterBarChartSegments
});
module.exports = __toCommonJS(MultiMeterBarChartSegments_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_MultiMeterBarChartSegment = require("./MultiMeterBarChartSegment.js");
var import_MeterBarLayoutSizeContext = require("../../shared/context/MeterBarLayoutSizeContext.js");
var import_get_segment_width = require("../../shared/utils/get-segment-width.js");
var import_selectors = require("../hooks/store/selectors.js");
const MultiMeterBarChartSegments = (props) => {
  const { max, min } = props;
  const { segmentWidth: width } = (0, import_MeterBarLayoutSizeContext.useMeterBarLayoutSize)();
  const segments = (0, import_selectors.useMultiMeterBarVisibleSegmentsState)();
  if (!segments || !segments.length) {
    return null;
  }
  const { segmentComponents } = segments.reduce(
    (acc, current, index) => {
      const segmentWidth = (0, import_get_segment_width.getSegmentWidth)(current.value, min, max, width);
      const segmentComponent = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_MultiMeterBarChartSegment.MultiMeterBarChartSegment,
        {
          previousSegmentsWidth: acc.usedWidth,
          segmentId: current.id,
          segmentIndex: index,
          segmentColor: current.color,
          segmentWidth
        },
        current.id
      );
      acc = {
        segmentComponents: [...acc.segmentComponents, segmentComponent],
        usedWidth: acc.usedWidth + segmentWidth
      };
      return acc;
    },
    {
      segmentComponents: [],
      usedWidth: 0
    }
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, { children: segmentComponents });
};
MultiMeterBarChartSegments["displayName"] = "MultiMeterBarChartSegments";
