// @dynatrace/strato-components/charts/meter-bar/multi-meter-bar/components/MultiMeterBarChartLegend.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MultiMeterBarChartLegend_exports = {};
__export(MultiMeterBarChartLegend_exports, {
  MultiMeterBarChartLegend: () => MultiMeterBarChartLegend
});
module.exports = __toCommonJS(MultiMeterBarChartLegend_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_ChartLegend = require("../../../core/components/legend/categorical/ChartLegend.js");
var import_LegendSizeContext = require("../../../core/components/legend/shared/context/LegendSizeContext.js");
var import_useTooltipReducer = require("../../../core/components/overlay-tooltip/useTooltipReducer.js");
var import_MeterBarChartLoadingContext = require("../../shared/context/MeterBarChartLoadingContext.js");
var import_MeterBarLayoutSizeContext = require("../../shared/context/MeterBarLayoutSizeContext.js");
var import_selectors = require("../hooks/store/selectors.js");
var import_useMultiMeterBarStore = require("../hooks/useMultiMeterBarStore.js");
const MultiMeterBarChartLegend = () => {
  const { segmentWidth: width } = (0, import_MeterBarLayoutSizeContext.useMeterBarLayoutSize)();
  const { setVisibleSegments, setHighlightedSegment } = (0, import_useMultiMeterBarStore.useMultiMeterBarStore)();
  const loading = (0, import_MeterBarChartLoadingContext.useMeterBarChartLoading)();
  const dispatch = (0, import_useTooltipReducer.useTooltipReducer)();
  const segments = (0, import_selectors.useMultiMeterBarSegmentsState)();
  const legendItems = segments.map((segment) => ({
    id: segment.id,
    color: {
      Default: segment.color
    },
    name: segment.name
  }));
  const onClickHandler = (visibleSegments) => {
    setVisibleSegments(visibleSegments);
    dispatch({ type: "RESET_TOOLTIP" })();
  };
  const onHoverHandler = (highlightedSegment) => {
    if (loading) {
      return;
    }
    setHighlightedSegment(highlightedSegment);
  };
  return (
    // the height value is set through the CSS style
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_LegendSizeContext.LegendSizeContext.Provider, { value: { width, height: 0 }, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_ChartLegend.ChartLegend,
      {
        items: legendItems,
        onClick: onClickHandler,
        onHover: onHoverHandler,
        interactionsDisabled: loading,
        style: { height: "100% " }
      }
    ) })
  );
};
MultiMeterBarChartLegend["displayName"] = "MultiMeterBarChartLegend";
