// @dynatrace/strato-components/charts/pie/components/PieChartLabel.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var PieChartLabel_exports = {};
__export(PieChartLabel_exports, {
  PieChartLabel: () => PieChartLabel
});
module.exports = __toCommonJS(PieChartLabel_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
var import_constants = require("../../core/constants.js");
var import_default = require("../types/default.js");
const PieChartLabel = ({ x, y, text, isHighlighted }) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
  "text",
  {
    fill: import_colors.default.Text.Neutral.Default,
    x,
    y,
    fontSize: import_default.LABEL_FONT_SIZE,
    textAnchor: "middle",
    pointerEvents: "none",
    opacity: isHighlighted ? 1 : import_constants.INACTIVE_FILL_OPACITY_PIE_CHART_LABEL,
    children: text
  }
);
