// @dynatrace/strato-components/charts/pie/components/PieChartInternal.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var PieChartInternal_exports = {};
__export(PieChartInternal_exports, {
  PieChartInternal: () => PieChartInternal
});
module.exports = __toCommonJS(PieChartInternal_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_PieChartLegend = require("./PieChartLegend/PieChartLegend.js");
var import_PieChartOverlays = require("./PieChartOverlays/PieChartOverlays.js");
var import_PieChartPlot = require("./PieChartPlot/PieChartPlot.js");
var import_useBehavioralTrackingProps = require("../../../core/hooks/useBehavioralTrackingProps.js");
var import_useMergeRefs = require("../../../core/hooks/useMergeRefs.js");
var import_ChartLayout = require("../../core/components/chart-layout/ChartLayout.js");
var import_useAutoLegendRefresh = require("../../core/components/chart-layout/hooks/useAutoLegendRefresh.js");
var import_OverlayContext = require("../../core/components/overlay-tooltip/contexts/OverlayContext.js");
var import_TooltipContext = require("../../core/components/overlay-tooltip/contexts/TooltipContext.js");
var import_constants = require("../../core/constants.js");
var import_sanitize_legend_ratio = require("../../core/utils/sanitize-legend-ratio.js");
var import_ChartLegendContext = require("../context/ChartLegendContext.js");
var import_PieChartRawDataContext = require("../context/PieChartRawDataContext.js");
var import_messages = require("../messages.js");
var import_PieChartSlicesDataProvider = require("../providers/PieChartSlicesDataProvider.js");
var import_default = require("../types/default.js");
const PieChartInternal = (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    height,
    width,
    isEmpty,
    emptyState,
    errorState,
    loading = false,
    className,
    style,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "data-testid": dataTestId,
    "aria-label": ariaLabel,
    "data-dt-component": dataDtComponent,
    role
  } = props;
  const [behavioralTrackingProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(props);
  const {
    hidden: legendHidden,
    position: legendPosition,
    // eslint-disable-next-line @typescript-eslint/no-deprecated
    ratio,
    initialRatio,
    onRatioChange
  } = (0, import_ChartLegendContext.useChartLegend)();
  const legendRatio = initialRatio ?? ratio;
  const plotAreaRef = (0, import_react.useRef)(null);
  const mergedRef = (0, import_useMergeRefs.useMergeRefs)([forwardedRef, plotAreaRef]);
  const intl = (0, import_react_intl.useIntl)();
  const chartData = (0, import_PieChartRawDataContext.usePieChartRawData)();
  const isInvalidDataset = chartData.slices?.reduce((acc, slice) => acc + slice.value, 0) <= 0;
  const invalidState = !isEmpty && isInvalidDataset ? emptyState ?? intl.formatMessage(import_messages.PIE_CHART_MESSAGES.invalidDataset) : emptyState;
  const overrideRatio = (orientation, width2, height2, ratio2, legendContainerDimensions) => {
    const isHorizontal = orientation === import_constants.ORIENTATION.Horizontal;
    if (!isHorizontal) {
      return legendContainerDimensions;
    }
    const sanitizedRatio = (0, import_sanitize_legend_ratio.sanitizeLegendRatio)(ratio2);
    const legendHeight = isHorizontal ? height2 : sanitizedRatio * height2 * 0.01;
    let legendWidth = width2;
    if (isHorizontal) {
      legendWidth = ratio2 === "auto" ? width2 - height2 : width2 * sanitizedRatio * 0.01;
    }
    return {
      width: legendWidth,
      height: legendHeight
    };
  };
  const chartLayoutRef = (0, import_useAutoLegendRefresh.useAutoLegendRefresh)(chartData);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayContext.OverlayProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TooltipContext.TooltipProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ChartLayout.ChartLayout,
    {
      ...behavioralTrackingProps,
      ref: chartLayoutRef,
      chartWidth: width ?? "100%",
      chartHeight: height ?? import_default.DEFAULT_SIZE,
      isEmpty: isEmpty || isInvalidDataset,
      emptyState: isEmpty ? emptyState : emptyState ?? invalidState,
      errorState,
      showLoader: loading,
      className,
      style: { padding: `0 ${import_default.DEFAULT_CONTAINER_PADDING}`, ...style },
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "data-dt-component": dataDtComponent,
      "data-testid": dataTestId,
      "aria-label": ariaLabel,
      role,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_ChartLayout.ChartLayout.Provider, { provider: import_PieChartSlicesDataProvider.PieChartSlicesDataProvider, children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ChartLayout.ChartLayout.Graph, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_PieChartOverlays.PieChartOverlays, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_PieChartPlot.PieChartPlot, { ...props, ref: mergedRef }) }) }),
        !legendHidden && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_ChartLayout.ChartLayout.Legend,
          {
            position: legendPosition,
            initialRatio: legendRatio,
            ratioBoundaries: import_default.PIE_CATEGORICAL_LEGEND_SIZES,
            overrideRatio,
            onResize: onRatioChange,
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_PieChartLegend.PieChartLegend, {})
          }
        )
      ] })
    }
  ) }) });
});
PieChartInternal.displayName = "PieChartInternal";
