// @dynatrace/strato-components/esm/tables/DataTable/features/Pagination/PaginationInternal.js
import { jsx, jsxs } from "react/jsx-runtime";
import { useIntl } from "react-intl";
import Breakpoints from "@dynatrace/strato-design-tokens/breakpoints-js";
import {
  ChevronLeftSmallIcon,
  ChevronRightSmallIcon
} from "@dynatrace/strato-icons";
import { Button } from "../../../../buttons/button/Button.js";
import { useId } from "../../../../core/hooks/useId.js";
import { Select } from "../../../../forms/select/Select.js";
import { Flex } from "../../../../layouts/flex/Flex.js";
import { useBreakpoint } from "../../../../layouts/hooks/useBreakpoint.js";
import { Text } from "../../../../typography/text/Text.js";
import { useDataTable } from "../../contexts/DataTableContext.js";
import { dataTableMessages } from "../../messages.js";
const PaginationNavigation = ({
  pageSize,
  pageCount,
  pageIndex,
  isPrevPageDisabled,
  isNextPageDisabled,
  isServerSidePaginated,
  totalRowsCount,
  onPageIndexChange
}) => {
  const pageIndexSelectId = useId();
  const intl = useIntl();
  const smallScreen = useBreakpoint(`(max-width: ${Breakpoints.Max.Mobile}px)`);
  const totalPages = isServerSidePaginated && totalRowsCount !== void 0 ? Math.ceil(totalRowsCount / pageSize) : pageCount;
  const pageCountArray = Array.from(
    { length: totalPages },
    (_, i) => i
  );
  const hasMultiplePages = totalPages > 1;
  const shouldShowPageInfo = !isServerSidePaginated || isServerSidePaginated && totalPages !== -1;
  const totalPagesCountText = hasMultiplePages ? intl.formatMessage(dataTableMessages.paginationTotalPages, { totalPages }) : intl.formatMessage(dataTableMessages.paginationSinglePageInfo);
  const handlePageIndexSelect = (newPageIndex) => newPageIndex !== null && onPageIndexChange(newPageIndex, "user");
  const handlePreviousPage = () => {
    const prevPageIndex = Math.max(0, pageIndex - 1);
    onPageIndexChange(prevPageIndex, "user");
  };
  const handleNextPage = () => onPageIndexChange(pageIndex + 1, "user");
  return /* @__PURE__ */ jsxs(Flex, { gap: 16, justifySelf: "flex-end", children: [
    shouldShowPageInfo && /* @__PURE__ */ jsxs(Flex, { flexWrap: "nowrap", alignItems: "center", children: [
      /* @__PURE__ */ jsx(Text, { as: "p", children: intl.formatMessage(dataTableMessages.paginationPageSelection) }),
      hasMultiplePages && /* @__PURE__ */ jsx(
        Select,
        {
          id: pageIndexSelectId,
          value: pageIndex,
          onChange: handlePageIndexSelect,
          "aria-label": intl.formatMessage(dataTableMessages.selectPage),
          "data-dt-name": "PaginationPageIndex",
          children: /* @__PURE__ */ jsx(Select.Content, { children: pageCountArray.length === 0 ? /* @__PURE__ */ jsx(Select.Option, { id: 0, value: 0, children: 0 }, "pageIndex-0") : pageCountArray.map((pageIndexOption) => /* @__PURE__ */ jsx(
            Select.Option,
            {
              id: pageIndexOption,
              value: pageIndexOption,
              children: pageIndexOption + 1
            },
            `pageIndex-${pageIndexOption}`
          )) })
        }
      ),
      (!hasMultiplePages || !smallScreen) && /* @__PURE__ */ jsx(Text, { as: "span", textStyle: "base", style: { whiteSpace: "nowrap" }, children: totalPagesCountText })
    ] }),
    (!smallScreen || isServerSidePaginated) && /* @__PURE__ */ jsxs(Flex, { gap: 2, children: [
      /* @__PURE__ */ jsx(
        Button,
        {
          onClick: handlePreviousPage,
          disabled: isPrevPageDisabled,
          "aria-label": intl.formatMessage(
            dataTableMessages.paginationPrevPage
          ),
          variant: "default",
          "data-dt-name": "PaginationPreviousPage",
          children: /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(ChevronLeftSmallIcon, {}) })
        }
      ),
      /* @__PURE__ */ jsx(
        Button,
        {
          onClick: handleNextPage,
          disabled: isNextPageDisabled,
          "aria-label": intl.formatMessage(
            dataTableMessages.paginationNextPage
          ),
          variant: "default",
          "data-dt-name": "PaginationNextPage",
          children: /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(ChevronRightSmallIcon, {}) })
        }
      )
    ] })
  ] });
};
function DataTablePaginationInternal(props) {
  const table = useDataTable();
  const intl = useIntl();
  const pageSizeSelectId = useId();
  const paginationState = table.getState().pagination;
  const pageSize = paginationState.pageSize;
  const onPageSizeChange = table.setPageSize;
  const pageSizeOptions = table.options.availablePaginationOptions ?? [];
  const hasPageSizeOptions = pageSizeOptions.length > 1;
  const { manualPagination } = table.options;
  const { enablePreviousPage, enableNextPage } = paginationState;
  const isServerSidePaginated = enableNextPage !== void 0 || enablePreviousPage !== void 0;
  const isPrevPageDisabled = manualPagination ? !enablePreviousPage : !table.getCanPreviousPage();
  const isNextPageDisabled = manualPagination ? !enableNextPage : !table.getCanNextPage();
  const {
    id,
    "data-testid": dataTestId,
    className: consumerClassName,
    style: consumerStyle,
    ...remainingProps
  } = props;
  return /* @__PURE__ */ jsxs(
    Flex,
    {
      id,
      "data-testid": dataTestId,
      className: consumerClassName,
      style: consumerStyle,
      justifyContent: hasPageSizeOptions ? "space-between" : "flex-end",
      paddingTop: 8,
      role: "navigation",
      flexWrap: "wrap",
      ...remainingProps,
      children: [
        hasPageSizeOptions && /* @__PURE__ */ jsxs(Flex, { alignItems: "center", children: [
          /* @__PURE__ */ jsx(
            Select,
            {
              id: pageSizeSelectId,
              value: pageSize,
              required: true,
              onChange: (newPageSize) => newPageSize !== null && onPageSizeChange?.(newPageSize),
              "aria-label": intl.formatMessage(dataTableMessages.selectPageSize),
              "data-dt-name": "PaginationPageSize",
              children: /* @__PURE__ */ jsx(Select.Content, { children: pageSizeOptions.map((pageSize2) => /* @__PURE__ */ jsx(
                Select.Option,
                {
                  id: pageSize2,
                  value: pageSize2,
                  children: `${pageSize2}`
                },
                `${pageSize2}`
              )) })
            }
          ),
          /* @__PURE__ */ jsx(Text, { as: "span", style: { whiteSpace: "nowrap" }, children: intl.formatMessage(dataTableMessages.paginationRowsPerPage) })
        ] }),
        /* @__PURE__ */ jsx(
          PaginationNavigation,
          {
            pageSize,
            pageCount: table.getPageCount(),
            pageIndex: paginationState.pageIndex,
            isPrevPageDisabled,
            isNextPageDisabled,
            isServerSidePaginated,
            totalRowsCount: paginationState.totalRowsCount,
            onPageIndexChange: table.setPageIndex
          }
        )
      ]
    }
  );
}
export {
  DataTablePaginationInternal
};
