// @dynatrace/strato-components/esm/tables/DataTable/features/Pagination/Pagination.js
import { functionalUpdate } from "@tanstack/react-table";
import {
  Children
} from "react";
import { isElement } from "react-is";
import { getPaginationRowModel } from "./pagination-row-model.js";
import { useControllableState } from "../../../../core/hooks/useControllableState.js";
import { useDataTableConfigElement } from "../../../hooks/useDataTableConfig.js";
import {
  DATA_TABLE_DEFAULT_PAGE_SIZE_OPTIONS,
  DATA_TABLE_DEFAULT_PAGINATION_PAGE_INDEX,
  DATA_TABLE_DEFAULT_PAGE_SIZE_OPTION
} from "../../constants.js";
function DataTablePagination(props) {
  return null;
}
function _isDataTablePagination(child) {
  return isElement(child) && child.type === DataTablePagination;
}
function getDefaultPageSize(pageSizeOptions, defaultPageSize) {
  if (typeof defaultPageSize === "number") {
    return defaultPageSize;
  }
  if (Array.isArray(pageSizeOptions) && pageSizeOptions[0]) {
    return pageSizeOptions[0];
  }
  return DATA_TABLE_DEFAULT_PAGE_SIZE_OPTION;
}
const DataTablePaginationFeature = {
  createTable(table) {
    table.setPagination = (updater, trigger) => {
      const safeUpdater = (old) => {
        const newState = functionalUpdate(
          updater,
          old
        );
        return newState;
      };
      return table.options.onPaginationChange?.(safeUpdater, trigger);
    };
    table.setPageIndex = (updater, trigger) => {
      table.setPagination((old) => {
        let pageIndex = functionalUpdate(updater, old.pageIndex);
        const maxPageIndex = typeof table.options.pageCount === "undefined" || table.options.pageCount === -1 ? Number.MAX_SAFE_INTEGER : table.options.pageCount - 1;
        pageIndex = Math.max(0, Math.min(pageIndex, maxPageIndex));
        return {
          ...old,
          pageIndex
        };
      }, trigger);
    };
    table.resetPageIndex = (defaultState) => {
      table.setPageIndex(
        defaultState ? DATA_TABLE_DEFAULT_PAGINATION_PAGE_INDEX : table.initialState?.pagination?.pageIndex ?? DATA_TABLE_DEFAULT_PAGINATION_PAGE_INDEX,
        "internal"
      );
    };
    table.getPrePaginationRowModel = () => table.getRowOrderRowModel();
  }
};
function usePagination(props, options) {
  const configPagination = useDataTableConfigElement("pagination");
  const paginationElement = Children.toArray(props.children).find(
    _isDataTablePagination
  );
  options.enablePagination = false;
  const paginationSizeOptions = paginationElement?.props.pageSizeOptions ?? [
    ...DATA_TABLE_DEFAULT_PAGE_SIZE_OPTIONS
  ];
  const defaultPaginationSizeOption = getDefaultPageSize(
    paginationElement?.props.pageSizeOptions,
    paginationElement?.props.defaultPageSize ?? configPagination?.defaultPageSize
  );
  const [pageSize, setPageSize] = useControllableState({
    value: paginationElement?.props.pageSize,
    defaultValue: defaultPaginationSizeOption,
    onChange: paginationElement?.props.onPageSizeChange
  });
  const defaultPaginationIndexOption = paginationElement?.props.defaultPageIndex ?? configPagination?.defaultPageIndex ?? DATA_TABLE_DEFAULT_PAGINATION_PAGE_INDEX;
  const [pageIndex, setPageIndex] = useControllableState({
    value: paginationElement?.props.pageIndex,
    defaultValue: defaultPaginationIndexOption
  });
  if (!paginationElement) {
    return;
  }
  options.enablePagination = true;
  options.availablePaginationOptions = paginationSizeOptions;
  options.controlledPageIndex = paginationElement.props.pageIndex !== void 0;
  if (paginationElement.props.autoResetPageIndex === false) {
    options.autoResetPageIndex = paginationElement.props.autoResetPageIndex;
  }
  if (paginationElement.props.enableNextPage !== void 0 && paginationElement.props.enablePreviousPage !== void 0) {
    options.manualPagination = true;
    options.pageCount = -1;
  } else {
    options.manualPagination = false;
    options.getPaginationRowModel = getPaginationRowModel();
  }
  options.state = {
    ...options.state,
    pagination: {
      pageSize,
      pageIndex,
      enablePreviousPage: paginationElement.props.enablePreviousPage,
      enableNextPage: paginationElement.props.enableNextPage,
      totalRowsCount: paginationElement.props.totalRowsCount
    }
  };
  options.onPaginationChange = (updater, trigger = "user") => {
    const nextState = functionalUpdate(
      updater,
      options.state?.pagination ?? {
        pageSize,
        pageIndex
      }
    );
    if (nextState?.pageSize !== pageSize && nextState?.pageSize !== void 0) {
      setPageSize(nextState.pageSize);
      setPageIndex(DATA_TABLE_DEFAULT_PAGINATION_PAGE_INDEX);
      paginationElement?.props.onPageIndexChange?.(
        DATA_TABLE_DEFAULT_PAGINATION_PAGE_INDEX,
        "internal"
      );
    } else if (nextState?.pageIndex !== pageIndex && nextState?.pageIndex !== void 0) {
      setPageIndex(nextState.pageIndex);
      paginationElement?.props.onPageIndexChange?.(
        nextState.pageIndex,
        trigger
      );
    }
  };
}
export {
  DataTablePagination,
  DataTablePaginationFeature,
  _isDataTablePagination,
  usePagination
};
