// @dynatrace/strato-components/esm/forms/date-time-picker/DateTimePicker.js
import { jsx } from "react/jsx-runtime";
import {
  forwardRef,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState
} from "react";
import { DateTimePickerRoot } from "./DateTimePickerRoot.js";
import { useCallbackRef } from "../../core/hooks/useCallbackRef.js";
import { useMergeRefs } from "../../core/hooks/useMergeRefs.js";
import { getDateTimePickerActions } from "../../core/time/timeframeStore/actions/dateTimePicker.js";
import { createTimeframeStore } from "../../core/time/timeframeStore/createTimeframeStore.js";
import { DateTimePickerStoreProvider } from "../../core/time/timeframeStore/timeframeStoreProvider.js";
import { useDateTimeMessages } from "../../core/time/useDateTimeMessages.js";
import { useMinMax, useMinMaxStoreSync } from "../../core/time/useMinMax.js";
import { validate } from "../../core/time/validation.js";
import { toCalendarValue } from "../../forms/forms-core/time/helpers.js";
import {
  removePrecisionFromTimeValue,
  getTimeValue
} from "../../forms/forms-core/time/timeParser.js";
import { getTimeOffsetFromInitialValue } from "../calendar/utils/helpers.js";
import { useFormFieldMessagesContext } from "../forms-core/hooks/useFormFieldMessagesContext.js";
import { useSyncedFormControlProps } from "../forms-core/hooks/useSyncedFormControlProps.js";
import {
  GRAIL_MAX_DATE,
  GRAIL_MIN_DATE
} from "../forms-core/time/date-time.js";
function useSyncPropsAndState(store, memoedValue, isControlled, precision, formatString, required, formFieldMessagesDispatch, dateTimePickerRef) {
  useEffect(() => {
    isControlled && store.setState((state) => {
      if (memoedValue?.absoluteDate === "") {
        return state;
      }
      const error = validate(memoedValue ?? null, state.min, state.max, {
        formMessageDispatch: formFieldMessagesDispatch,
        formerError: state.calendarError,
        precision: state.precision,
        showError: false,
        checkOrder: true,
        ref: dateTimePickerRef,
        expectedFormat: state.formatString,
        valueMode: "single"
      });
      const value = removePrecisionFromTimeValue(
        memoedValue ?? null,
        state.precision
      );
      const timeOffset = getTimeOffsetFromInitialValue(
        value,
        "single",
        state.min,
        state.max,
        state.precision
      );
      return {
        ...state,
        dateTimePickerValue: value,
        calendarValue: toCalendarValue(value, state.precision),
        calendarError: error,
        dateTimePickerError: error,
        timeOffset
      };
    });
  }, [
    formFieldMessagesDispatch,
    isControlled,
    memoedValue,
    store,
    dateTimePickerRef
  ]);
  useEffect(() => {
    store.setState((state) => {
      const value = removePrecisionFromTimeValue(
        state.dateTimePickerValue,
        precision
      );
      const error = validate(value, state.min, state.max, {
        formMessageDispatch: formFieldMessagesDispatch,
        formerError: state.dateTimePickerError,
        precision,
        showError: false,
        checkOrder: true,
        ref: dateTimePickerRef,
        expectedFormat: formatString,
        valueMode: "single",
        required
      });
      return {
        ...state,
        formatString,
        precision,
        dateTimePickerValue: value,
        calendarValue: toCalendarValue(value, precision),
        required,
        calendarError: error,
        dateTimePickerError: error
      };
    });
  }, [
    formatString,
    precision,
    store,
    required,
    formFieldMessagesDispatch,
    dateTimePickerRef
  ]);
}
const DateTimePicker = forwardRef(
  ({
    value: propValue,
    defaultValue: propDefaultValue,
    onChange,
    min: propMin = GRAIL_MIN_DATE,
    max: propMax = GRAIL_MAX_DATE,
    precision: propPrecision = "minutes",
    disabled: propDisabled,
    required: propRequired,
    readOnly: propReadOnly,
    id: propId,
    "aria-disabled": propAriaDisabled,
    type,
    ...remainingProps
  }, forwardedRef) => {
    const [isControlled] = useState(propValue !== void 0);
    const onChangeProp = useCallbackRef(onChange);
    const precision = type === "date" ? "day" : propPrecision;
    const memoedValue = useMemo(
      () => getTimeValue(propValue, precision) ?? null,
      [precision, propValue]
    );
    const memoedDefaultValue = useMemo(
      () => getTimeValue(propDefaultValue, precision),
      /** We're ignoring updates to defaultValue as it only considers the initial state. */
      // eslint-disable-next-line react-hooks/exhaustive-deps
      []
    );
    const { min, max } = useMinMax(propMin, propMax, precision);
    const calendarRef = useRef(null);
    const dateTimeInputRef = useRef(null);
    const dateTimePickerRef = useRef(null);
    const { dispatch: formFieldMessagesDispatch } = useFormFieldMessagesContext();
    const { disabled, required, id, ariaDisabled, readOnly } = useSyncedFormControlProps({
      propDisabled,
      propRequired,
      propId,
      propAriaDisabled,
      propReadOnly
    });
    const { formatString } = useDateTimeMessages({ precision });
    const [initialState] = useState(() => {
      const _value = isControlled ? memoedValue : memoedDefaultValue;
      const value = _value ?? null;
      const timeOffset = getTimeOffsetFromInitialValue(
        value,
        "single",
        min,
        max,
        precision
      );
      const error = validate(value, min, max, {
        formMessageDispatch: formFieldMessagesDispatch,
        precision,
        showError: false,
        checkOrder: true,
        ref: dateTimePickerRef,
        expectedFormat: formatString,
        valueMode: "single"
      });
      const calendarValue = toCalendarValue(value, precision);
      return {
        calendarValue,
        dateTimePickerValue: value,
        min,
        max,
        precision,
        calendarRef,
        dateTimeInputRef,
        dateTimePickerRef,
        calendarError: error,
        dateTimePickerError: error,
        timeOffset
      };
    });
    const [store] = useState(
      () => createTimeframeStore(
        initialState,
        getDateTimePickerActions(
          formFieldMessagesDispatch,
          onChangeProp,
          isControlled
        )
      )
    );
    useSyncPropsAndState(
      store,
      memoedValue,
      isControlled,
      precision,
      formatString,
      required,
      formFieldMessagesDispatch,
      dateTimePickerRef
    );
    useMinMaxStoreSync(
      store,
      min,
      max,
      useCallback(
        (state, min2, max2) => {
          const error = validate(state.dateTimePickerValue, min2, max2, {
            formMessageDispatch: formFieldMessagesDispatch,
            formerError: state.calendarError,
            precision: state.precision,
            showError: false,
            checkOrder: true,
            ref: dateTimePickerRef,
            expectedFormat: state.formatString,
            valueMode: "single"
          });
          return { calendarError: error, dateTimePickerError: error };
        },
        [formFieldMessagesDispatch]
      )
    );
    const mergedRef = useMergeRefs([dateTimePickerRef, forwardedRef]);
    return /* @__PURE__ */ jsx(DateTimePickerStoreProvider, { store, children: /* @__PURE__ */ jsx(
      DateTimePickerRoot,
      {
        ...remainingProps,
        id,
        ref: mergedRef,
        "aria-disabled": ariaDisabled,
        disabled,
        required,
        readOnly,
        type
      }
    ) });
  }
);
export {
  DateTimePicker
};
