// @dynatrace/strato-components/esm/forms/date-time-picker/DateSelector.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { forwardRef, useImperativeHandle, useState } from "react";
import { useIntl } from "react-intl";
import { CalendarIcon } from "@dynatrace/strato-icons";
import { Button } from "../../buttons/button/Button.js";
import { useMergeRefs } from "../../core/hooks/useMergeRefs.js";
import { isVisualRefreshEnabled } from "../../core/utils/is-visual-refresh-enabled.js";
import { useOverlayWithTrigger } from "../../overlays/hooks/useOverlayWithTrigger.js";
import { Overlay } from "../../overlays/overlay/Overlay.js";
import { globalSprinkles } from "../../styles/safe-sprinkles.js";
import { CalendarRoot } from "../calendar/CalendarRoot.js";
import { FormFieldMessageContext } from "../forms-core/contexts/FormFieldMessageContext.js";
import { useFormFieldMessagesContext } from "../forms-core/hooks/useFormFieldMessagesContext.js";
const DateSelector = forwardRef(
  (props, forwardedRef) => {
    const {
      disabled,
      error,
      readOnly,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "aria-disabled": ariaDisabledProp,
      ...remainingProps
    } = props;
    const [open, setOpen] = useState(false);
    const {
      targetRef,
      triggerRef,
      overlayRef,
      overlayState,
      overlayProps,
      overlayTriggerProps
    } = useOverlayWithTrigger({
      placement: "bottom-end",
      offset: 8,
      trapFocus: false,
      shouldCloseOnBlur: false,
      open,
      onOpenChange: setOpen,
      // the overlay should not open when disabled or readOnly
      disabled: disabled || readOnly
    });
    const rootRef = useMergeRefs([triggerRef, targetRef]);
    const intl = useIntl();
    const calendarInputLabel = intl.formatMessage({
      defaultMessage: "Calendar for picking a date",
      id: "qpf+aqg3hXoviypu",
      description: "Aria label for the DateTimePicker calendar input used for form validation."
    });
    useImperativeHandle(
      forwardedRef,
      () => ({
        open: () => setOpen(true),
        close: () => setOpen(false)
      }),
      []
    );
    const { state: formFieldState } = useFormFieldMessagesContext();
    return /* @__PURE__ */ jsxs(Fragment, { children: [
      /* @__PURE__ */ jsx(
        Button,
        {
          ref: rootRef,
          "data-testid": dataTestId,
          className: consumerClassName,
          "aria-disabled": ariaDisabledProp,
          style: consumerStyle,
          ...remainingProps,
          ...overlayTriggerProps,
          size: "condensed",
          color: error ? "critical" : "neutral",
          variant: isVisualRefreshEnabled() ? "emphasized" : "default",
          "data-dt-name": "ShowCalendar",
          children: /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(CalendarIcon, {}) })
        }
      ),
      /* @__PURE__ */ jsx(
        Overlay,
        {
          ref: overlayRef,
          isOpen: overlayState.isOpen,
          className: globalSprinkles({
            paddingLeft: 16,
            paddingTop: 16,
            paddingRight: 16,
            paddingBottom: 16
          }),
          overlayProps,
          "data-dtrum-mask": dataDtrumMask,
          "data-dtrum-allow": dataDtrumAllow,
          "aria-label": intl.formatMessage({
            defaultMessage: "Calendar overlay for picking a date",
            id: "A9Pdb5T48IHfEauq",
            description: "Aria label for the date time picker overlay."
          }),
          children: /* @__PURE__ */ jsx(
            FormFieldMessageContext.Provider,
            {
              value: {
                state: {
                  customMessages: [],
                  formControlMessages: [],
                  showError: formFieldState.showError
                }
              },
              children: /* @__PURE__ */ jsx(
                CalendarRoot,
                {
                  "aria-label": calendarInputLabel,
                  "aria-disabled": ariaDisabledProp
                }
              )
            }
          )
        }
      )
    ] });
  }
);
export {
  DateSelector
};
