// @dynatrace/strato-components/esm/charts/core/components/grid/Grid.js
import { jsx, jsxs } from "react/jsx-runtime";
import { Group } from "@visx/group";
import { GridColumns } from "../../../core/components/grid/GridColumns.js";
import { GridRows } from "../../../core/components/grid/GridRows.js";
import {
  getVerticalNumTicks,
  getHorizontalNumTicks
} from "../../../core/utils/ticks.js";
import { isBandScale } from "../../utils/detect-scale.js";
const DEFAULTS = {
  HIGHLIGHT_ZERO_LINE: {
    horizontal: true,
    vertical: true
  }
};
function Grid({
  width,
  height,
  xScale,
  yScale,
  highlightZeroLine = DEFAULTS.HIGHLIGHT_ZERO_LINE,
  xTicks,
  yTicks,
  xCategoryOffset,
  yCategoryOffset
}) {
  const numTicksRows = getVerticalNumTicks(height);
  const autoNumTicks = getHorizontalNumTicks(width);
  const xNumTicks = xCategoryOffset !== void 0 && isBandScale(xScale) ? xScale.domain().length : autoNumTicks;
  const yNumTicks = yCategoryOffset !== void 0 && isBandScale(yScale) ? yScale.domain().length : numTicksRows;
  return /* @__PURE__ */ jsxs(Group, { children: [
    /* @__PURE__ */ jsx(
      GridColumns,
      {
        width,
        height,
        scale: xScale,
        numTicks: xNumTicks,
        highlightZeroLine: highlightZeroLine.vertical,
        ticks: xTicks,
        offset: xCategoryOffset
      }
    ),
    /* @__PURE__ */ jsx(
      GridRows,
      {
        width,
        height,
        scale: yScale,
        numTicks: yNumTicks,
        highlightZeroLine: highlightZeroLine.horizontal,
        ticks: yTicks,
        offset: yCategoryOffset
      }
    )
  ] });
}
export {
  Grid
};
