// @dynatrace/strato-components/esm/charts/core/components/annotations/components/tooltip/AnnotationsOverlayTooltip.js
import { jsx } from "react/jsx-runtime";
import { useEffect, useRef } from "react";
import { AnnotationsCustomizedTooltip } from "./AnnotationsCustomizedTooltip.js";
import { AnnotationsDefaultTooltipContent } from "./AnnotationsDefaultTooltipContent.js";
import { useAnnotationsTooltipConfig } from "./hooks/useAnnotationsTooltipConfig.js";
import { OverlayTooltip } from "../../../overlay-tooltip/OverlayTooltip.js";
import { useTooltipReducer } from "../../../overlay-tooltip/useTooltipReducer.js";
import { emptyData } from "../../../overlay-tooltip/utils.js";
import { useAnnotationsTooltipState } from "../../hooks/store/selectors.js";
import {
  useSetStateSelection,
  useSetStateTooltip
} from "../../hooks/store/state-setters.js";
import { useMarkersInteractions } from "../../hooks/useMarkersInteractions.js";
import { initialState } from "../../providers/AnnotationsDataInitialState.js";
const buildSentinelData = () => ({
  geometry: "rect",
  series: [
    {
      name: "",
      value: "",
      color: ""
    }
  ]
});
const AnnotationsOverlayTooltip = (props) => {
  const { tooltipType, truncationMode = "end", maxLines = 1 } = props;
  const annotationsState = useAnnotationsTooltipState();
  const { pinned, data, visibility, position } = annotationsState;
  const dispatch = useTooltipReducer();
  const updateAnnotationsTooltipState = useSetStateTooltip();
  const updateAnnotationsSelectionState = useSetStateSelection();
  const onMarkersInteraction = useMarkersInteractions();
  const config = useAnnotationsTooltipConfig();
  const hidden = config?.hidden === true;
  const hasData = !!data && data.length > 0 && visibility;
  const prevPinnedRef = useRef(false);
  useEffect(() => {
    const wasPinned = prevPinnedRef.current;
    prevPinnedRef.current = pinned;
    if (wasPinned && !pinned) {
      dispatch({
        type: "SET_PINNED",
        payload: {
          pinned: false,
          data: emptyData()
        }
      })();
    }
    if (hidden || !hasData || !position) {
      dispatch({ type: "SET_OUT_BOUNDS" })();
      return;
    }
    if (pinned) {
      dispatch({
        type: "SET_PINNED",
        payload: {
          pinned: true,
          absoluteX: position.x,
          absoluteY: position.y,
          data: buildSentinelData()
        }
      })();
    } else {
      dispatch({
        type: "SET_POSITION",
        payload: {
          absoluteX: position.x,
          absoluteY: position.y,
          data: buildSentinelData()
        }
      })();
    }
  }, [hidden, hasData, pinned, position?.x, position?.y]);
  useEffect(() => {
    return () => {
      dispatch({ type: "RESET_TOOLTIP" })();
      updateAnnotationsTooltipState(initialState.tooltip);
      updateAnnotationsSelectionState(initialState.selection);
      onMarkersInteraction?.(initialState.selection);
    };
  }, []);
  useEffect(() => {
    const handleEscape = (event) => {
      if (event.key !== "Escape") {
        return;
      }
      updateAnnotationsTooltipState(initialState.tooltip);
      updateAnnotationsSelectionState(initialState.selection);
      onMarkersInteraction?.(initialState.selection);
    };
    window.addEventListener("keyup", handleEscape);
    return () => {
      window.removeEventListener("keyup", handleEscape);
    };
  }, []);
  if (hidden || !hasData) {
    return null;
  }
  const legacyTemplate = config?.template ? /* @__PURE__ */ jsx(
    AnnotationsCustomizedTooltip,
    {
      tooltipTemplate: config.template,
      data
    }
  ) : /* @__PURE__ */ jsx(
    AnnotationsDefaultTooltipContent,
    {
      data,
      tooltipType,
      truncationMode,
      maxLines
    }
  );
  return /* @__PURE__ */ jsx(OverlayTooltip, { legacyTemplate });
};
AnnotationsOverlayTooltip.displayName = "AnnotationsOverlayTooltip";
export {
  AnnotationsOverlayTooltip
};
