// @dynatrace/strato-components/esm/charts/core/components/annotations/components/tooltip/AnnotationsDefaultTooltipContent.js
import { jsx, jsxs } from "react/jsx-runtime";
import { Fragment } from "react";
import { useIntl } from "react-intl";
import { formatDate } from "@dynatrace-sdk/units";
import { useAnnotationsTooltipFormatter } from "./hooks/useAnnotationsTooltipFormatter.js";
import { getSymbol } from "./utils/get-symbol.js";
import { Divider } from "../../../../../../layouts/divider/Divider.js";
import { Footer } from "../../../overlay-tooltip/components/Footer.js";
import { Header } from "../../../overlay-tooltip/components/header/Header.js";
import { SeriesItem } from "../../../overlay-tooltip/components/series/SeriesItem.js";
import { ExternalConfigContext } from "../../../overlay-tooltip/contexts/ExternalConfigContext.js";
import { OVERLAY_TOOLTIP_MESSAGES } from "../../../overlay-tooltip/messages.js";
import { useSelector } from "../../../overlay-tooltip/selectors.js";
import {
  annotationsGroupsScrollContainer,
  divider,
  seriesGridContainer
} from "../../../overlay-tooltip/tooltip.sty.js";
import { useAnnotationsActions } from "../../hooks/useAnnotationsActions.js";
import { deserializeMarkerRange } from "../../utils/serialize-marker-range.js";
const BASE_ACTIONS_CONFIG = /* @__PURE__ */ new Map([
  ["name", { field: "valueLabel" }]
]);
const buildExternalConfig = (truncationMode) => ({
  seriesHeader: null,
  displayHeaderDivider: false,
  symbolPosition: "left",
  truncationMode,
  defaultActionsConfig: BASE_ACTIONS_CONFIG,
  variant: "single"
});
const X_VALUE_LABEL = "X-value";
const AnnotationsDefaultTooltipContent = (props) => {
  const { data, tooltipType, truncationMode, maxLines } = props;
  const formatter = useAnnotationsTooltipFormatter();
  const annotationsActions = useAnnotationsActions();
  const config = buildExternalConfig(truncationMode);
  const intl = useIntl();
  const unpinnedPlaceholderText = intl.formatMessage(
    OVERLAY_TOOLTIP_MESSAGES.unpinnedPlaceholderMessage
  );
  const pinned = useSelector((state) => state.pinned);
  return /* @__PURE__ */ jsxs(ExternalConfigContext.Provider, { value: config, children: [
    /* @__PURE__ */ jsx("div", { className: annotationsGroupsScrollContainer, children: data.map((annotationGroup, groupIndex) => {
      const [from, to] = deserializeMarkerRange(annotationGroup.key);
      const isRange = to > from;
      let header1;
      let showRangeDuration = false;
      let duration;
      if (tooltipType === "numerical" && isRange) {
        header1 = [formatter(from), formatter(to)];
        duration = String(Math.abs(to - from));
      } else if (tooltipType === "numerical") {
        header1 = formatter(from);
      } else if (isRange) {
        header1 = [new Date(from), new Date(to)];
        showRangeDuration = true;
      } else {
        header1 = formatDate(new Date(from), {
          day: "2-digit",
          month: "short",
          year: "numeric",
          hour: "2-digit",
          minute: "2-digit"
        });
      }
      return /* @__PURE__ */ jsxs(Fragment, { children: [
        groupIndex > 0 && /* @__PURE__ */ jsx(Divider, { orientation: "horizontal", className: divider }),
        /* @__PURE__ */ jsx(
          Header,
          {
            header1,
            label: X_VALUE_LABEL,
            showRangeDuration1: showRangeDuration,
            duration1: duration,
            maxLines
          }
        ),
        /* @__PURE__ */ jsx("div", { className: seriesGridContainer, children: annotationGroup.markers.map((marker) => {
          const icon = getSymbol(marker);
          const {
            id,
            symbol,
            title: markerTitle,
            description: markerDescription,
            start,
            end,
            color,
            priority,
            hidden: markerHidden,
            ...extraFields
          } = marker;
          const payload = {
            ...extraFields,
            start,
            end,
            title: markerTitle,
            description: markerDescription,
            symbol
          };
          const actions = annotationsActions?.(payload);
          return /* @__PURE__ */ jsx(
            SeriesItem,
            {
              name: markerTitle ?? "",
              description: markerDescription,
              value: "",
              icon,
              multiLine: maxLines > 1,
              maxLines,
              actions: actions ? () => actions : void 0
            },
            id
          );
        }) })
      ] }, annotationGroup.key);
    }) }),
    !pinned && /* @__PURE__ */ jsx(
      Footer,
      {
        actions: [],
        unpinnedPlaceholderText
      }
    )
  ] });
};
AnnotationsDefaultTooltipContent.displayName = "AnnotationsDefaultTooltipContent";
export {
  AnnotationsDefaultTooltipContent
};
