// @dynatrace/strato-components/esm/layouts/split-layout/slots.js
import {
  Children,
  forwardRef,
  isValidElement
} from "react";
const PanelSlot = forwardRef(
  (_props, _ref) => null
);
const ContentSlot = forwardRef(
  (_props, _ref) => null
);
const ResizeHandleSlot = forwardRef(
  (_props, _ref) => null
);
function isPanelSlot(element) {
  return element.type === PanelSlot;
}
function isContentSlot(element) {
  return element.type === ContentSlot;
}
function isResizeHandleSlot(element) {
  return element.type === ResizeHandleSlot;
}
function extractSlotsProps(children) {
  let panel = null;
  let content = null;
  let resizeHandle = null;
  Children.forEach(children, (child) => {
    if (!isValidElement(child)) {
      return;
    }
    if (isPanelSlot(child)) {
      if (panel) {
        throw new Error("SplitLayout: only one Panel slot is allowed");
      }
      const {
        minWidth,
        maxWidth,
        defaultWidth,
        position,
        resizable,
        children: children2,
        onResize,
        label,
        ...rest
      } = child.props;
      panel = {
        children: children2,
        config: {
          minWidth: minWidth ?? 200,
          maxWidth: maxWidth ?? Infinity,
          defaultWidth: defaultWidth ?? 200,
          position: position ?? "start",
          resizable: resizable ?? true,
          onResize,
          label
        },
        domProps: rest,
        ref: "ref" in child ? child.ref : void 0
      };
    } else if (isContentSlot(child)) {
      if (content) {
        throw new Error("SplitLayout: only one Content slot is allowed");
      }
      const { children: children2, minWidth, ...rest } = child.props;
      content = {
        children: children2,
        config: {
          minWidth: minWidth ?? 250
        },
        domProps: rest,
        ref: "ref" in child ? child.ref : void 0
      };
    } else if (isResizeHandleSlot(child)) {
      if (resizeHandle) {
        throw new Error("SplitLayout: only one ResizeHandle slot is allowed");
      }
      const { children: children2, ...rest } = child.props;
      resizeHandle = {
        children: children2,
        domProps: rest,
        ref: "ref" in child ? child.ref : void 0
      };
    }
  });
  if (!panel) {
    throw new Error("Panel slot is required");
  }
  return { panel, content, resizeHandle };
}
export {
  ContentSlot,
  PanelSlot,
  ResizeHandleSlot,
  extractSlotsProps
};
