// @dynatrace/strato-components/esm/layouts/split-layout/split-sizing-utils.js
import {
  clamp,
  resolveWidthPx,
  toCssLength
} from "./width-utils.js";
function clampPanelWidth(split, containerWidth, config) {
  const { panelMin, panelMax, contentMin, handleWidth } = config;
  const minPx = resolveWidthPx(panelMin, containerWidth, 0);
  const maxPx = resolveWidthPx(panelMax, containerWidth, Infinity);
  const contentMinPx = resolveWidthPx(contentMin, containerWidth, 0);
  let extrinsicMax = Infinity;
  if (containerWidth !== 0) {
    extrinsicMax = containerWidth - contentMinPx - handleWidth;
  }
  const finalMax = Math.max(minPx, Math.min(maxPx, extrinsicMax));
  return clamp(minPx, split, finalMax);
}
function computeLayout(splitPx, containerWidth, config) {
  const { panelMin, panelMax, contentMin, handleWidth } = config;
  const panelWidth = clampPanelWidth(splitPx, containerWidth, config);
  const panelMinPx = resolveWidthPx(panelMin, containerWidth, 0);
  const panelMaxPx = resolveWidthPx(panelMax, containerWidth, Infinity);
  const contentMinPx = resolveWidthPx(contentMin, containerWidth, 0);
  const contentMaxPx = Math.max(0, containerWidth - contentMinPx - handleWidth);
  const contentWidth = clamp(
    contentMinPx,
    containerWidth - handleWidth - panelWidth,
    contentMaxPx
  );
  const contentWidthCss = `max(${toCssLength(contentMin)}, 100% - ${panelWidth ? panelWidth + handleWidth : 0}px)`;
  return {
    panel: { width: panelWidth, minPx: panelMinPx, maxPx: panelMaxPx },
    content: {
      width: contentWidth,
      widthCss: contentWidthCss,
      minPx: contentMinPx,
      maxPx: contentMaxPx
    }
  };
}
export {
  clampPanelWidth,
  computeLayout
};
