// @dynatrace/strato-components/esm/layouts/split-layout/SplitLayout.js
import { jsx, jsxs } from "react/jsx-runtime";
import { useGesture } from "@use-gesture/react";
import { clsx } from "clsx";
import {
  createContext,
  forwardRef,
  memo,
  useContext,
  useRef
} from "react";
import {
  ContentSlot,
  extractSlotsProps,
  PanelSlot,
  ResizeHandleSlot
} from "./slots.js";
import * as styles from "./SplitLayout.sty.js";
import { useSplitSizing } from "./useSplitSizing.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useId } from "../../core/hooks/useId.js";
import { useMergeRefs } from "../../core/hooks/useMergeRefs.js";
import { useFocusRing } from "../../core/styles/useFocusRing.js";
const PanelShell = memo(({ children }) => {
  return children;
});
const ContentShell = memo(({ children }) => {
  return children;
});
const ResizeHandleShell = memo(({ children }) => {
  return children;
});
const PanelColumn = forwardRef(
  (props, forwardedRef) => {
    const { children, className, style, ...rest } = props;
    return /* @__PURE__ */ jsx(
      "div",
      {
        ...rest,
        ref: forwardedRef,
        className: clsx(styles.column, className),
        style,
        children
      }
    );
  }
);
const ContentColumn = forwardRef(
  (props, forwardedRef) => {
    const { children, className, style, ...rest } = props;
    return /* @__PURE__ */ jsx(
      "div",
      {
        ...rest,
        ref: forwardedRef,
        className: clsx(styles.column, className),
        style,
        children
      }
    );
  }
);
const ResizeHandle = forwardRef((props, forwardedRef) => {
  const { onDrag, onDragStart, onDragEnd, panelId, ariaLabel, ariaLabelledBy } = useContext(ActionsContext);
  const ref = useRef(null);
  const mergedRefs = useMergeRefs([ref, forwardedRef]);
  const { focusProps, focusClassName } = useFocusRing({
    focusWithin: true,
    ignoreModality: true
  });
  useGesture(
    {
      onDragStart,
      onDragEnd,
      onDrag: (state) => {
        onDrag(state.movement[0]);
      }
    },
    { target: ref }
  );
  return /* @__PURE__ */ jsx(
    "div",
    {
      ...props,
      ...focusProps,
      role: "separator",
      ref: mergedRefs,
      className: clsx(styles.resizeHandle, focusClassName),
      tabIndex: 0,
      "aria-controls": panelId,
      "aria-label": ariaLabelledBy ? void 0 : ariaLabel,
      "aria-labelledby": ariaLabelledBy
    }
  );
});
const ActionsContext = createContext({
  onDragStart: () => void 0,
  onDrag: () => void 0,
  onDragEnd: () => void 0,
  panelId: "",
  ariaLabel: void 0,
  ariaLabelledBy: void 0
});
const SplitLayoutRoot = memo(
  forwardRef((props, forwardedRef) => {
    const { children, className: rootClassname, ...rootProps } = props;
    const [behavioralTrackingProps, restRootProps] = useBehavioralTrackingProps(rootProps);
    const slots = extractSlotsProps(children);
    const rootRef = useRef(null);
    const rootMergedRef = useMergeRefs([forwardedRef, rootRef]);
    const generatedPanelId = useId();
    const panelId = slots.panel.domProps.id ?? generatedPanelId;
    const panelRef = useRef(null);
    const contentRef = useRef(null);
    const panelMergedRef = useMergeRefs([panelRef, slots.panel.ref]);
    const contentMergedRef = useMergeRefs([contentRef, slots.content?.ref]);
    const sizing = useSplitSizing({
      panel: {
        min: slots.panel.config.minWidth,
        max: slots.panel.config.maxWidth,
        default: slots.panel.config.defaultWidth,
        position: slots.panel.config.position,
        resizable: slots.panel.config.resizable,
        onResize: slots.panel.config.onResize
      },
      content: {
        min: slots.content?.config.minWidth ?? 0
      },
      refs: {
        root: rootRef,
        panel: panelRef,
        content: contentRef
      }
    });
    const isDefaultPosition = slots.panel.config.position === "start";
    const activeSizing = isDefaultPosition ? sizing.panel : sizing.content;
    const panelColumn = /* @__PURE__ */ jsx(
      PanelColumn,
      {
        ...slots.panel.domProps,
        id: panelId,
        style: { ...slots.panel.domProps.style, width: sizing.panel.widthCss },
        ref: panelMergedRef,
        children: /* @__PURE__ */ jsx(PanelShell, { children: slots.panel.children })
      }
    );
    const contentColumn = slots.content ? /* @__PURE__ */ jsx(
      ContentColumn,
      {
        ...slots.content.domProps,
        style: {
          ...slots.content.domProps.style,
          width: sizing.content.widthCss,
          // content needs an explicit minWidth when it is the leading column
          // (end position) to prevent it from collapsing before the resize handle
          minWidth: isDefaultPosition ? void 0 : slots.content.config.minWidth
        },
        ref: contentMergedRef,
        children: /* @__PURE__ */ jsx(ContentShell, { children: slots.content.children })
      }
    ) : null;
    const resizeHandleDecorations = slots.resizeHandle?.children ? /* @__PURE__ */ jsx("div", { className: styles.handleDecorations, children: /* @__PURE__ */ jsx(ResizeHandleShell, { children: slots.resizeHandle.children }) }) : null;
    const resizeHandle = slots.panel.config.resizable ? /* @__PURE__ */ jsxs(
      "div",
      {
        ...slots.resizeHandle?.domProps,
        className: clsx(
          styles.handleColumn,
          slots.resizeHandle?.domProps?.className
        ),
        ref: slots.resizeHandle?.ref,
        children: [
          /* @__PURE__ */ jsx(
            ResizeHandle,
            {
              "aria-valuemin": activeSizing.minPx,
              "aria-valuemax": isFinite(activeSizing.maxPx) ? activeSizing.maxPx : void 0,
              "aria-valuenow": activeSizing.width,
              "aria-valuetext": `${activeSizing.width}px`,
              "aria-orientation": "vertical"
            }
          ),
          resizeHandleDecorations
        ]
      }
    ) : null;
    return /* @__PURE__ */ jsx(
      "div",
      {
        ...restRootProps,
        ...behavioralTrackingProps,
        className: clsx(styles.layout, rootClassname),
        ref: rootMergedRef,
        children: /* @__PURE__ */ jsxs(
          ActionsContext.Provider,
          {
            value: {
              ...sizing.dragActions,
              panelId,
              ariaLabel: slots.resizeHandle?.domProps["aria-label"] ?? slots.panel.config.label,
              ariaLabelledBy: slots.resizeHandle?.domProps["aria-labelledby"]
            },
            children: [
              isDefaultPosition ? panelColumn : contentColumn,
              resizeHandle,
              isDefaultPosition ? contentColumn : panelColumn
            ]
          }
        )
      }
    );
  })
);
SplitLayoutRoot.displayName = "SplitLayout.Root";
const SplitLayout = Object.assign(SplitLayoutRoot, {
  Panel: PanelSlot,
  Content: ContentSlot,
  ResizeHandle: ResizeHandleSlot
});
export {
  SplitLayout
};
