// @dynatrace/strato-components/esm/layouts/split-layout/useSplitSizing.js
import { throttle } from "lodash-es";
import {
  useCallback,
  useLayoutEffect,
  useMemo,
  useRef,
  useState
} from "react";
import useResizeObserver from "use-resize-observer";
import { clampPanelWidth, computeLayout } from "./split-sizing-utils.js";
import { convertToPx, toCssLength } from "./width-utils.js";
function useSplitSizing(options) {
  const { panel, content, refs } = options;
  const [containerWidth, setContainerWidth] = useState(0);
  const [splitPx, setSplitPx] = useState(() => {
    if (typeof panel.default === "number") {
      return panel.default;
    }
    return -1;
  });
  const splitPxRef = useRef(splitPx);
  const isDraggingRef = useRef(false);
  const hasBeenManuallyResizedRef = useRef(false);
  const dragStartPx = useRef(null);
  if (!isDraggingRef.current) {
    splitPxRef.current = splitPx;
  }
  const handleWidth = panel.resizable ? 8 : 0;
  const contentMinCss = toCssLength(content.min);
  const sizingConfig = useMemo(
    () => ({
      panelMin: panel.min,
      panelMax: panel.max,
      contentMin: content.min,
      handleWidth
    }),
    [panel.min, panel.max, content.min, handleWidth]
  );
  const getClampedSplit = useCallback(
    (split, container) => clampPanelWidth(split, container, sizingConfig),
    [sizingConfig]
  );
  useLayoutEffect(
    function initializeSplit() {
      if (containerWidth === 0) {
        return;
      }
      if (typeof panel.default === "number") {
        if (splitPx !== -1) {
          return;
        }
        setSplitPx(panel.default);
        return;
      }
      if (hasBeenManuallyResizedRef.current) {
        return;
      }
      setSplitPx(convertToPx(panel.default, containerWidth));
    },
    [containerWidth, panel.default, splitPx]
  );
  const KEEP_SQUISHED_SIZES = false;
  const throttledSync = useMemo(
    () => throttle((width) => {
      const expected = getClampedSplit(splitPxRef.current, width);
      if (splitPxRef.current !== expected) {
        setSplitPx(expected);
      }
    }, 100),
    [getClampedSplit]
  );
  useResizeObserver({
    ref: refs.root,
    onResize: (size) => {
      if (!size.width) {
        return;
      }
      setContainerWidth(size.width);
      if (KEEP_SQUISHED_SIZES) {
        throttledSync(size.width);
      }
    }
  });
  const { position: panelPosition, onResize: onPanelResize } = panel;
  const dragActions = useMemo(
    () => ({
      onDragStart: () => {
        isDraggingRef.current = true;
        const containerW = refs.root.current?.offsetWidth ?? 0;
        const clamped = getClampedSplit(splitPxRef.current, containerW);
        splitPxRef.current = clamped;
        setSplitPx(clamped);
        dragStartPx.current = clamped;
      },
      onDrag: (mx) => {
        const containerW = refs.root.current?.offsetWidth ?? 0;
        if (dragStartPx.current === null || containerW === 0) {
          return;
        }
        const sign = panelPosition === "start" ? 1 : -1;
        const next = dragStartPx.current + mx * sign;
        const clamped = getClampedSplit(next, containerW);
        splitPxRef.current = clamped;
        if (refs.panel.current) {
          refs.panel.current.style.width = `${clamped}px`;
        }
        if (refs.content.current) {
          refs.content.current.style.width = `max(${contentMinCss}, 100% - ${clamped + handleWidth}px)`;
        }
      },
      onDragEnd: () => {
        isDraggingRef.current = false;
        hasBeenManuallyResizedRef.current = true;
        dragStartPx.current = null;
        const rootWidth = refs.root.current?.offsetWidth ?? 0;
        const clamped = getClampedSplit(splitPxRef.current, rootWidth);
        splitPxRef.current = clamped;
        setSplitPx(clamped);
        onPanelResize?.(clamped);
      }
    }),
    [
      getClampedSplit,
      panelPosition,
      onPanelResize,
      handleWidth,
      contentMinCss,
      refs
    ]
  );
  const layout = computeLayout(splitPx, containerWidth, sizingConfig);
  const isUnresolved = splitPx === -1;
  const panelWidthCss = isUnresolved ? toCssLength(panel.default) : `${layout.panel.width}px`;
  const contentWidthCss = isUnresolved ? `max(${toCssLength(content.min)}, 100% - ${toCssLength(
    panel.default
  )} - ${handleWidth}px)` : layout.content.widthCss;
  return {
    ...layout,
    panel: { ...layout.panel, widthCss: panelWidthCss },
    content: { ...layout.content, widthCss: contentWidthCss },
    handleWidth,
    dragActions
  };
}
export {
  useSplitSizing
};
