// @dynatrace/strato-components/esm/charts/xy-chart/processing-strategies/strategies/area-series.js
import { SeriesBaseStrategy } from "./base-series.js";
import { GappableSeriesHelper } from "./gappable-series.js";
import { StackableSeriesHelper } from "./stackable-series.js";
import { isValidPointsDisplay } from "../../../core/components/canvas/utils/is-valid-points-display.js";
import { safeGet } from "../../../core/utils/safe-get.js";
import { DEFAULT_POINTS_DISPLAY } from "../../constants/defaults.js";
import { isAreaSeries } from "../../utils/type-guards.js";
import { getAreaDatapoints } from "../utils/access-datapoint/get-area-datapoints.js";
import {
  getChartAreaVariant,
  getChartOrientation,
  getDatumValue
} from "../utils/access-datapoint/get-datapoints.js";
import { getCalculatedLinearPointForArea } from "../utils/cast-data/get-calculated-linear-point.js";
import { getLinearColor } from "../utils/coloring/get-linear-color-palette.js";
import { isAreaSeriesStacked } from "../utils/stacking/is-area-series-stacked.js";
import {
  getBaselineCoordinate,
  getCombinedAxesOrientationKey,
  isStackable
} from "../utils/stacking/stack-datapoints.js";
class AreaSeriesStrategy extends SeriesBaseStrategy {
  stackHelper;
  gapHelper;
  /**
   * This is needed for special cases such as isolated points in value representation in area charts.
   */
  combinedAxesOrientationKeyToSeriesCount;
  /** This set contains series Id that has been preprocessed at least once */
  preprocessedSeriesIds;
  constructor(initializer) {
    super(initializer);
    this.stackHelper = new StackableSeriesHelper();
    this.gapHelper = new GappableSeriesHelper();
    this.combinedAxesOrientationKeyToSeriesCount = /* @__PURE__ */ new Map();
    this.preprocessedSeriesIds = /* @__PURE__ */ new Set();
  }
  canHandle(seriesProps) {
    return isAreaSeries(seriesProps);
  }
  getSeriesId(datum, seriesProps) {
    return getDatumValue(
      datum,
      seriesProps.seriesIdAccessor
    )?.toString();
  }
  createSeries(seriesId, seriesProps, xAxis, yAxis) {
    return {
      id: seriesId,
      type: seriesProps.type,
      data: [],
      xAxisId: xAxis.id,
      yAxisId: yAxis.id
    };
  }
  createSeriesMetadata(datum, seriesProps, xAxis, yAxis, seriesId, context) {
    const { intl, seriesIndexes } = context;
    const { nameAccessor: seriesNameAccessor, curve } = seriesProps;
    const seriesIndex = this.getOrCreateSeriesIndex(seriesId, seriesIndexes);
    const seriesName = this.getSeriesName(intl, seriesIndex, void 0, {
      seriesNameAccessor,
      datum
    });
    const color = this.getColor(
      datum,
      seriesProps,
      context.colorPalette,
      seriesName,
      seriesIndex
    );
    const chartOrientation = getChartOrientation(seriesProps);
    const orientation = chartOrientation === "horizontal" ? "horizontal" : "vertical";
    const variant = getChartAreaVariant(
      chartOrientation,
      seriesProps
    );
    return {
      xAxis,
      yAxis,
      ...color,
      name: seriesName,
      orientation,
      curve,
      variant,
      pointsDisplay: seriesProps.pointsDisplay && isValidPointsDisplay(seriesProps.pointsDisplay) ? seriesProps.pointsDisplay : DEFAULT_POINTS_DISPLAY,
      actions: seriesProps.actions
    };
  }
  getCalculationPoint(datapoint, xAxis, yAxis, chartOrientation, variant) {
    return getCalculatedLinearPointForArea(
      datapoint,
      xAxis,
      yAxis,
      chartOrientation,
      variant
    );
  }
  preprocessImpl(data, seriesProps, xAxis, yAxis, context) {
    const chartOrientation = getChartOrientation(seriesProps);
    if (chartOrientation === void 0 || chartOrientation === "both") {
      return;
    }
    const areaVariant = getChartAreaVariant(chartOrientation, seriesProps);
    if (areaVariant === "band") {
      return;
    }
    const canStack = isStackable("stack", chartOrientation, xAxis, yAxis);
    if (!canStack) {
      return;
    }
    const filteredSeriesIds = context?.filteredSeriesIds;
    for (const datum of data) {
      const datapoint = this.getDatapoint(datum, seriesProps, xAxis, yAxis);
      const variant = getChartAreaVariant(chartOrientation, seriesProps);
      if (datapoint) {
        const seriesId = this.getSeriesId(datum, seriesProps);
        const isSeriesVisible = filteredSeriesIds === void 0 || filteredSeriesIds.has(seriesId);
        if (!isSeriesVisible) {
          continue;
        }
        this.registerSeriesCountOnPreprocessing(
          seriesId,
          xAxis.id,
          yAxis.id,
          chartOrientation
        );
        const preCalcPoint = this.getCalculationPoint(
          datapoint,
          xAxis,
          yAxis,
          chartOrientation,
          variant
        );
        this.stackHelper.processStackDomain(
          preCalcPoint,
          chartOrientation,
          xAxis,
          yAxis,
          seriesId
        );
      }
    }
  }
  processImpl(datum, seriesProps, xAxis, yAxis, datapoint, context) {
    const { seriesMap, seriesToMetadata } = context;
    const seriesId = this.getSeriesId(datum, seriesProps);
    const isSeriesVisible = context.filteredSeriesIds === void 0 || context.filteredSeriesIds.has(seriesId);
    const chartOrientation = getChartOrientation(seriesProps);
    if (!chartOrientation) {
      return "continue";
    }
    const variant = getChartAreaVariant(chartOrientation, seriesProps);
    const preCalcPoint = this.getCalculationPoint(
      datapoint,
      xAxis,
      yAxis,
      chartOrientation,
      variant
    );
    let calcDatapoint;
    let datapointMetadata;
    let shouldContinue;
    if (isSeriesVisible) {
      ({ calcDatapoint, datapointMetadata, shouldContinue } = this.handleStacking(
        preCalcPoint,
        xAxis,
        yAxis,
        chartOrientation,
        variant
      ));
    } else {
      calcDatapoint = preCalcPoint;
      datapointMetadata = {
        isGap: false,
        baseline: getBaselineCoordinate(chartOrientation)
      };
      shouldContinue = false;
    }
    if (shouldContinue) {
      return "continue";
    }
    let currentSeries = seriesMap.get(seriesId);
    if (!currentSeries) {
      currentSeries = this.createSeries(seriesId, seriesProps, xAxis, yAxis);
      this.registerSeries(currentSeries, seriesId, context);
      const metadata = this.createSeriesMetadata(
        datum,
        seriesProps,
        xAxis,
        yAxis,
        seriesId,
        context
      );
      seriesToMetadata.set(currentSeries, {
        ...metadata,
        isVisible: isSeriesVisible,
        isStacked: chartOrientation !== "both" && isAreaSeriesStacked(
          getCombinedAxesOrientationKey(xAxis.id, yAxis.id, chartOrientation),
          this.combinedAxesOrientationKeyToSeriesCount,
          variant
        )
      });
    }
    const currentSeriesMetadata = safeGet(seriesToMetadata, currentSeries);
    if (isSeriesVisible) {
      context.axisWithSeries.add(xAxis.id);
      context.axisWithSeries.add(yAxis.id);
      const gappedDatapoint = this.gapHelper.handleGapDetection(
        datapoint,
        currentSeries,
        xAxis,
        yAxis
      );
      if (gappedDatapoint) {
        this.registerDatapoint(
          gappedDatapoint,
          { isGap: true },
          currentSeries,
          context
        );
      }
    }
    this.registerDatapoint(
      datapoint,
      datapointMetadata,
      currentSeries,
      context
    );
    this.updateAxisDomains(
      calcDatapoint,
      xAxis,
      yAxis,
      context.axisIdToScaleDomain,
      isSeriesVisible
    );
    if (isSeriesVisible) {
      this.addValue({
        series: currentSeries,
        datapoint,
        color: currentSeriesMetadata?.color
      });
    }
  }
  getColor(datum, seriesProps, configColorPalette, seriesName, seriesIndex) {
    return getLinearColor(
      datum,
      seriesProps,
      configColorPalette,
      seriesName,
      seriesIndex
    );
  }
  getDatapoint(datum, seriesProps, xAxis, yAxis) {
    return getAreaDatapoints(datum, seriesProps, xAxis, yAxis);
  }
  handleStacking(datapoint, xAxis, yAxis, chartOrientation, areaVariant) {
    if (areaVariant === "band") {
      return {
        calcDatapoint: datapoint,
        datapointMetadata: {
          isGap: false,
          baseline: getBaselineCoordinate(chartOrientation)
        },
        shouldContinue: false
      };
    }
    let calcDatapoint;
    let shouldContinue = false;
    let datapointMetadata = {
      isGap: false,
      baseline: getBaselineCoordinate(chartOrientation)
    };
    const canStack = isStackable("stack", chartOrientation, xAxis, yAxis);
    if (canStack) {
      ({ calcDatapoint, datapointMetadata, shouldContinue } = this.stackHelper.processStackedDatapoint(
        datapoint,
        chartOrientation,
        xAxis,
        yAxis,
        datapointMetadata
      ));
    }
    return {
      calcDatapoint: calcDatapoint ?? datapoint,
      datapointMetadata,
      shouldContinue
    };
  }
  registerSeriesCountForAxisCombination(xAxisId, yAxisId, chartOrientation) {
    const axesCombinationKey = getCombinedAxesOrientationKey(
      xAxisId,
      yAxisId,
      chartOrientation
    );
    const count = this.combinedAxesOrientationKeyToSeriesCount.get(axesCombinationKey) ?? 0;
    this.combinedAxesOrientationKeyToSeriesCount.set(
      axesCombinationKey,
      count + 1
    );
  }
  registerSeriesCountOnPreprocessing(seriesId, xAxisId, yAxisId, chartOrientation) {
    if (!this.preprocessedSeriesIds.has(seriesId)) {
      this.registerSeriesCountForAxisCombination(
        xAxisId,
        yAxisId,
        chartOrientation
      );
      this.preprocessedSeriesIds.add(seriesId);
    }
  }
}
export {
  AreaSeriesStrategy
};
