// @dynatrace/strato-components/esm/charts/xy-chart/processing-strategies/strategies/bar-series.js
import { SeriesBaseStrategy } from "./base-series.js";
import { GroupableSeriesHelper } from "./groupable-series.js";
import { StackableSeriesHelper } from "./stackable-series.js";
import {
  BAR_VARIANT_SINGLE
} from "../../types/xy-chart.js";
import { isBarSeries } from "../../utils/type-guards.js";
import { getBarDatapoint } from "../utils/access-datapoint/get-bar-datapoints.js";
import {
  getChartOrientation,
  getDatumValue
} from "../utils/access-datapoint/get-datapoints.js";
import { getLinearColor } from "../utils/coloring/get-linear-color-palette.js";
import { getBarSeriesVariant } from "../utils/grouping/get-variant.js";
import { isGrouping } from "../utils/grouping/group-bar-datapoints.js";
import {
  getBaselineCoordinate,
  isStackable
} from "../utils/stacking/stack-datapoints.js";
class BarSeriesStrategy extends SeriesBaseStrategy {
  stackHelper;
  groupHelper;
  constructor(initializer) {
    super(initializer);
    this.stackHelper = new StackableSeriesHelper();
    this.groupHelper = new GroupableSeriesHelper();
  }
  get axisIdToDimension() {
    return this.groupHelper.axisIdToDimension;
  }
  canHandle(seriesProps) {
    return isBarSeries(seriesProps);
  }
  getSeriesId(datum, seriesProps) {
    return getDatumValue(
      datum,
      seriesProps.seriesIdAccessor
    )?.toString();
  }
  createSeries(seriesId, seriesProps, xAxis, yAxis) {
    return {
      id: seriesId,
      type: seriesProps.type,
      data: [],
      xAxisId: xAxis.id,
      yAxisId: yAxis.id,
      labelAccessor: seriesProps.labelAccessor
    };
  }
  createSeriesMetadata(datum, seriesProps, xAxis, yAxis, seriesId, context, variant) {
    const { intl, seriesIndexes } = context;
    const { nameAccessor: seriesNameAccessor } = seriesProps;
    const seriesIndex = this.getOrCreateSeriesIndex(seriesId, seriesIndexes);
    const seriesName = this.getSeriesName(intl, seriesIndex, void 0, {
      seriesNameAccessor,
      datum
    });
    const color = this.getColor(
      datum,
      seriesProps,
      context.colorPalette,
      seriesName,
      seriesIndex
    );
    return {
      xAxis,
      yAxis,
      ...color,
      name: seriesName,
      variant: variant ?? BAR_VARIANT_SINGLE,
      actions: seriesProps.actions,
      chartOrientation: getChartOrientation(seriesProps)
    };
  }
  handleStackingAndGrouping(datapoint, seriesProps, xAxis, yAxis, chartOrientation, seriesId, isSeriesVisible) {
    let shouldContinue = false;
    let datapointMetadata = {
      isGap: false,
      baseline: getBaselineCoordinate(chartOrientation)
    };
    const canStack = isStackable(
      seriesProps.variant,
      chartOrientation,
      xAxis,
      yAxis
    );
    const canGroup = isGrouping(
      seriesProps.variant,
      chartOrientation,
      xAxis,
      yAxis
    );
    let calcDatapoint;
    if (!isSeriesVisible) {
      return {
        calcDatapoint: calcDatapoint ?? datapoint,
        datapointMetadata,
        shouldContinue,
        variant: getBarSeriesVariant(canStack, canGroup)
      };
    }
    if (canStack) {
      ({ calcDatapoint, datapointMetadata, shouldContinue } = this.stackHelper.processStackedDatapoint(
        datapoint,
        chartOrientation,
        xAxis,
        yAxis,
        datapointMetadata
      ));
    }
    if (canGroup) {
      ({ datapointMetadata } = this.groupHelper.processGroupDatapoint(
        datapoint,
        chartOrientation,
        xAxis,
        yAxis,
        seriesId,
        datapointMetadata
      ));
    }
    return {
      calcDatapoint: calcDatapoint ?? datapoint,
      datapointMetadata,
      shouldContinue,
      variant: getBarSeriesVariant(canStack, canGroup)
    };
  }
  preprocessImpl(data, seriesProps, xAxis, yAxis, context) {
    const chartOrientation = getChartOrientation(seriesProps);
    if (chartOrientation === void 0 || chartOrientation === "both") {
      return;
    }
    const canStack = isStackable(
      seriesProps.variant,
      chartOrientation,
      xAxis,
      yAxis
    );
    const canGroup = isGrouping(
      seriesProps.variant,
      chartOrientation,
      xAxis,
      yAxis
    );
    if (!canStack && !canGroup) {
      return;
    }
    const filteredSeriesIds = context?.filteredSeriesIds;
    for (const datum of data) {
      const datapoint = this.getDatapoint(datum, seriesProps, xAxis, yAxis);
      if (datapoint) {
        const seriesId = this.getSeriesId(datum, seriesProps);
        const isSeriesVisible = filteredSeriesIds === void 0 || filteredSeriesIds.has(seriesId);
        if (!isSeriesVisible) {
          continue;
        }
        if (canStack) {
          this.stackHelper.processStackDomain(
            datapoint,
            chartOrientation,
            xAxis,
            yAxis,
            seriesId
          );
        }
        if (canGroup) {
          this.groupHelper.processGroupDomain(
            datapoint,
            chartOrientation,
            xAxis,
            yAxis,
            seriesId
          );
        }
      }
    }
  }
  processImpl(datum, seriesProps, xAxis, yAxis, datapoint, context) {
    const { seriesMap, seriesToMetadata } = context;
    const seriesId = this.getSeriesId(datum, seriesProps);
    const chartOrientation = getChartOrientation(seriesProps);
    if (chartOrientation === void 0) {
      return "continue";
    }
    const isSeriesVisible = context.filteredSeriesIds === void 0 || context.filteredSeriesIds.has(seriesId);
    const { calcDatapoint, datapointMetadata, shouldContinue, variant } = this.handleStackingAndGrouping(
      datapoint,
      seriesProps,
      xAxis,
      yAxis,
      chartOrientation,
      seriesId,
      isSeriesVisible
    );
    if (shouldContinue) {
      return "continue";
    }
    let currentSeries = seriesMap.get(seriesId);
    if (!currentSeries) {
      currentSeries = this.createSeries(seriesId, seriesProps, xAxis, yAxis);
      this.registerSeries(currentSeries, seriesId, context);
      const metadata = this.createSeriesMetadata(
        datum,
        seriesProps,
        xAxis,
        yAxis,
        seriesId,
        context,
        variant
      );
      seriesToMetadata.set(currentSeries, {
        ...metadata,
        isVisible: isSeriesVisible
      });
    }
    if (datapoint.label) {
      currentSeries.hasLabels = true;
    }
    this.registerDatapoint(
      datapoint,
      datapointMetadata,
      currentSeries,
      context
    );
    this.updateAxisDomains(
      calcDatapoint,
      xAxis,
      yAxis,
      context.axisIdToScaleDomain,
      isSeriesVisible
    );
    if (isSeriesVisible) {
      context.axisWithSeries.add(xAxis.id);
      context.axisWithSeries.add(yAxis.id);
      this.addValue({
        series: currentSeries,
        datapoint
      });
    }
  }
  getColor(datum, seriesProps, configColorPalette, seriesName, seriesIndex) {
    return getLinearColor(
      datum,
      seriesProps,
      configColorPalette,
      seriesName,
      seriesIndex
    );
  }
  getDatapoint(datum, seriesProps, xAxis, yAxis) {
    return getBarDatapoint(datum, seriesProps, xAxis, yAxis);
  }
}
export {
  BarSeriesStrategy
};
