// @dynatrace/strato-components/esm/charts/xy-chart/processing-strategies/strategies/base-series.js
import { isUndefined } from "lodash-es";
import { XYAxisTypeCategorical } from "../../types/axis.js";
import { calculateAxesDomains } from "../../utils/calculate-axes-domains.js";
import {
  getSeriesNameWithAccessor,
  getSeriesNameWithFallback
} from "../utils/series-naming/get-series-name.js";
class SeriesBaseStrategy {
  addValue;
  constructor(initializer) {
    this.addValue = initializer.addValue;
  }
  preprocess(data, seriesProps, xAxis, yAxis, context) {
    if (!this.canHandle(seriesProps)) {
      return;
    }
    return this.preprocessImpl(data, seriesProps, xAxis, yAxis, context);
  }
  process(datum, seriesProps, xAxis, yAxis, context) {
    if (!this.canHandle(seriesProps)) {
      return;
    }
    const datapoint = this.getDatapoint(datum, seriesProps, xAxis, yAxis);
    if (!datapoint) {
      return "continue";
    }
    return this.processImpl(
      datum,
      seriesProps,
      xAxis,
      yAxis,
      datapoint,
      context
    );
  }
  preprocessImpl(data, seriesProps, xAxis, yAxis, context) {
    return;
  }
  /**
   * Gets or creates the series index, adding to seriesIndexes if new.
   */
  getOrCreateSeriesIndex(seriesId, seriesIndexes) {
    if (!seriesIndexes.includes(seriesId)) {
      seriesIndexes.push(seriesId);
    }
    return seriesIndexes.indexOf(seriesId);
  }
  /**
   * Creates a new series object. Override in subclasses for custom series structure.
   */
  createSeries(seriesId, seriesProps, xAxis, yAxis) {
    return {
      id: seriesId,
      type: seriesProps.type,
      data: [],
      xAxisId: xAxis.id,
      yAxisId: yAxis.id
    };
  }
  /**
   * Registers a new series in the context maps.
   */
  registerSeries(currentSeries, seriesId, context) {
    context.seriesMap.set(seriesId, currentSeries);
    context.seriesWithDatapoints.push(currentSeries);
  }
  /**
   * Registers a datapoint to the series and updates metadata and domains.
   */
  registerDatapoint(datapoint, datapointMetadata, currentSeries, context) {
    currentSeries.data.push(datapoint);
    context.datapointMetadataMap.set(datapoint, datapointMetadata);
  }
  /**
   * Calculates and sets the axis domains.
   */
  updateAxisDomains(calcPoint, xAxis, yAxis, axisIdToScaleDomain, isSeriesVisible) {
    const { xDomain, yDomain } = calculateAxesDomains(
      xAxis,
      yAxis,
      calcPoint,
      axisIdToScaleDomain,
      isSeriesVisible
    );
    axisIdToScaleDomain.set(xAxis.id, xDomain);
    if (isSeriesVisible || yAxis.type === XYAxisTypeCategorical) {
      axisIdToScaleDomain.set(yAxis.id, yDomain);
    }
  }
  /**
   * Gets the calculation point for domain updates. Override for custom logic (e.g., time centering).
   */
  getCalculationPoint(datapoint, _xAxis, _yAxis) {
    return datapoint;
  }
  getSeriesName(intl, seriesIndex, seriesNameFallback, namingWithAccessor) {
    return isUndefined(namingWithAccessor) ? getSeriesNameWithFallback(intl, seriesIndex, seriesNameFallback) : getSeriesNameWithAccessor(
      intl,
      seriesIndex,
      seriesNameFallback,
      namingWithAccessor.seriesNameAccessor,
      namingWithAccessor.datum
    );
  }
}
export {
  SeriesBaseStrategy
};
