// @dynatrace/strato-components/esm/charts/xy-chart/processing-strategies/strategies/dot-series.js
import { SeriesBaseStrategy } from "./base-series.js";
import { safeGet } from "../../../core/utils/safe-get.js";
import { isDotSeries } from "../../utils/type-guards.js";
import {
  getDatumValue,
  getDotDatapoint
} from "../utils/access-datapoint/get-datapoints.js";
import { getCalculatedLinearPoint } from "../utils/cast-data/get-calculated-linear-point.js";
import { getLinearColor } from "../utils/coloring/get-linear-color-palette.js";
class DotSeriesStrategy extends SeriesBaseStrategy {
  canHandle(seriesProps) {
    return isDotSeries(seriesProps);
  }
  // can this method be un base?
  getSeriesId(datum, seriesProps) {
    const seriesIdAccessor = getDatumValue(
      datum,
      seriesProps.seriesIdAccessor
    )?.toString();
    return seriesIdAccessor ?? seriesProps.id;
  }
  createSeriesMetadata(datum, seriesProps, xAxis, yAxis, seriesId, context) {
    const { intl, seriesIndexes } = context;
    const { nameAccessor: seriesNameAccessor } = seriesProps;
    const seriesIndex = this.getOrCreateSeriesIndex(seriesId, seriesIndexes);
    const seriesName = this.getSeriesName(
      intl,
      seriesIndex,
      // APPDEV-15290 - Adjust after removing name prop from series
      // eslint-disable-next-line @typescript-eslint/no-deprecated
      seriesProps.name,
      {
        seriesNameAccessor,
        datum
      }
    );
    const color = this.getColor(
      datum,
      seriesProps,
      context.colorPalette,
      seriesName,
      seriesIndex
    );
    return {
      xAxis,
      yAxis,
      actions: seriesProps.actions,
      ...color,
      name: seriesName,
      shape: seriesProps.shape
    };
  }
  getCalculationPoint(datapoint, xAxis, yAxis) {
    return getCalculatedLinearPoint(datapoint, xAxis, yAxis);
  }
  processImpl(datum, seriesProps, xAxis, yAxis, datapoint, context) {
    const { seriesMap, seriesToMetadata } = context;
    const seriesId = this.getSeriesId(datum, seriesProps);
    const isSeriesVisible = context.filteredSeriesIds === void 0 || context.filteredSeriesIds.has(seriesId);
    let currentSeries = seriesMap.get(seriesId);
    if (!currentSeries) {
      currentSeries = this.createSeries(
        seriesId,
        seriesProps,
        xAxis,
        yAxis
      );
      this.registerSeries(currentSeries, seriesId, context);
      const metadata = this.createSeriesMetadata(
        datum,
        seriesProps,
        xAxis,
        yAxis,
        seriesId,
        context
      );
      seriesToMetadata.set(currentSeries, {
        ...metadata,
        isVisible: isSeriesVisible
      });
    }
    const datapointMetadata = { isGap: false };
    this.registerDatapoint(
      datapoint,
      datapointMetadata,
      currentSeries,
      context
    );
    const calcPoint = this.getCalculationPoint(datapoint, xAxis, yAxis);
    this.updateAxisDomains(
      calcPoint,
      xAxis,
      yAxis,
      context.axisIdToScaleDomain,
      isSeriesVisible
    );
    if (isSeriesVisible) {
      context.axisWithSeries.add(xAxis.id);
      context.axisWithSeries.add(yAxis.id);
      const currentSeriesMetadata = safeGet(seriesToMetadata, currentSeries);
      this.addValue({
        series: currentSeries,
        datapoint,
        color: currentSeriesMetadata?.color
      });
    }
  }
  getColor(datum, seriesProps, configColorPalette, seriesName, seriesIndex) {
    return getLinearColor(
      datum,
      seriesProps,
      configColorPalette,
      seriesName,
      seriesIndex
    );
  }
  getDatapoint(datum, seriesProps, xAxis, yAxis) {
    return getDotDatapoint(datum, seriesProps, xAxis, yAxis);
  }
}
export {
  DotSeriesStrategy
};
