// @dynatrace/strato-components/esm/charts/xy-chart/processing-strategies/strategies/line-series.js
import { SeriesBaseStrategy } from "./base-series.js";
import { GappableSeriesHelper } from "./gappable-series.js";
import { isValidPointsDisplay } from "../../../core/components/canvas/utils/is-valid-points-display.js";
import { safeGet } from "../../../core/utils/safe-get.js";
import { DEFAULT_POINTS_DISPLAY } from "../../constants/defaults.js";
import { isLineSeries } from "../../utils/type-guards.js";
import {
  getDatumValue,
  getLineDatapoint
} from "../utils/access-datapoint/get-datapoints.js";
import { getCalculatedLinearPoint } from "../utils/cast-data/get-calculated-linear-point.js";
import { getLinearColor } from "../utils/coloring/get-linear-color-palette.js";
class LineSeriesStrategy extends SeriesBaseStrategy {
  gapHelper;
  constructor(initializer) {
    super(initializer);
    this.gapHelper = new GappableSeriesHelper();
  }
  canHandle(seriesProps) {
    return isLineSeries(seriesProps);
  }
  getSeriesId(datum, seriesProps) {
    return getDatumValue(
      datum,
      seriesProps.seriesIdAccessor
    )?.toString();
  }
  createSeriesMetadata(datum, seriesProps, xAxis, yAxis, seriesId, context) {
    const { intl, seriesIndexes } = context;
    const { nameAccessor: seriesNameAccessor } = seriesProps;
    const seriesIndex = this.getOrCreateSeriesIndex(seriesId, seriesIndexes);
    const seriesName = this.getSeriesName(intl, seriesIndex, void 0, {
      seriesNameAccessor,
      datum
    });
    const color = this.getColor(
      datum,
      seriesProps,
      context.colorPalette,
      seriesName,
      seriesIndex
    );
    return {
      xAxis,
      yAxis,
      curve: seriesProps.curve,
      actions: seriesProps.actions,
      pointsDisplay: seriesProps.pointsDisplay && isValidPointsDisplay(seriesProps.pointsDisplay) ? seriesProps.pointsDisplay : DEFAULT_POINTS_DISPLAY,
      ...color,
      name: seriesName
    };
  }
  getCalculationPoint(datapoint, xAxis, yAxis) {
    return getCalculatedLinearPoint(datapoint, xAxis, yAxis);
  }
  processImpl(datum, seriesProps, xAxis, yAxis, datapoint, context) {
    const { seriesMap, seriesToMetadata } = context;
    const seriesId = this.getSeriesId(datum, seriesProps);
    const isSeriesVisible = context.filteredSeriesIds === void 0 || context.filteredSeriesIds.has(seriesId);
    let currentSeries = seriesMap.get(seriesId);
    if (!currentSeries) {
      currentSeries = this.createSeries(
        seriesId,
        seriesProps,
        xAxis,
        yAxis
      );
      this.registerSeries(currentSeries, seriesId, context);
      const metadata = this.createSeriesMetadata(
        datum,
        seriesProps,
        xAxis,
        yAxis,
        seriesId,
        context
      );
      seriesToMetadata.set(currentSeries, {
        ...metadata,
        isVisible: isSeriesVisible
      });
    }
    const datapointMetadata = { isGap: false };
    if (isSeriesVisible) {
      const gappedDatapoint = this.gapHelper.handleGapDetection(
        datapoint,
        currentSeries,
        xAxis,
        yAxis
      );
      if (gappedDatapoint) {
        this.registerDatapoint(
          gappedDatapoint,
          { isGap: true },
          currentSeries,
          context
        );
      }
    }
    this.registerDatapoint(
      datapoint,
      datapointMetadata,
      currentSeries,
      context
    );
    const calcPoint = this.getCalculationPoint(datapoint, xAxis, yAxis);
    this.updateAxisDomains(
      calcPoint,
      xAxis,
      yAxis,
      context.axisIdToScaleDomain,
      isSeriesVisible
    );
    if (isSeriesVisible) {
      context.axisWithSeries.add(xAxis.id);
      context.axisWithSeries.add(yAxis.id);
      const currentSeriesMetadata = safeGet(seriesToMetadata, currentSeries);
      this.addValue({
        series: currentSeries,
        datapoint,
        color: currentSeriesMetadata?.color
      });
    }
  }
  getColor(datum, seriesProps, configColorPalette, seriesName, seriesIndex) {
    return getLinearColor(
      datum,
      seriesProps,
      configColorPalette,
      seriesName,
      seriesIndex
    );
  }
  getDatapoint(datum, seriesProps, xAxis, yAxis) {
    return getLineDatapoint(datum, seriesProps, xAxis, yAxis);
  }
}
export {
  LineSeriesStrategy
};
