// @dynatrace/strato-components/esm/charts/xy-chart/processing-strategies/strategies/rect-series.js
import { SeriesBaseStrategy } from "./base-series.js";
import { safeGet } from "../../../core/utils/safe-get.js";
import { DEFAULT_COLOR_PALETTE_RECT } from "../../constants/defaults.js";
import {
  RANGE_LEGEND_TYPE
} from "../../types/xy-chart-internals.js";
import {
  isRectSeries,
  isStringOrValidNumber
} from "../../utils/type-guards.js";
import { getRectDatapoint } from "../utils/access-datapoint/get-datapoints.js";
import { getLegendType } from "../utils/cast-data/get-legend-type.js";
import { getValidColorPalette } from "../utils/coloring/get-valid-color-palette.js";
import { getValueScaleDomain } from "../utils/scale/get-value-scale-domain.js";
class RectSeriesStrategy extends SeriesBaseStrategy {
  seriesIdToValueScaleDomain;
  constructor(initializer) {
    super(initializer);
    this.seriesIdToValueScaleDomain = /* @__PURE__ */ new Map();
  }
  canHandle(seriesProps) {
    return isRectSeries(seriesProps);
  }
  getSeriesId(_datum, seriesProps) {
    return seriesProps.id;
  }
  createSeries(seriesId, seriesProps, xAxis, yAxis) {
    return {
      id: seriesId,
      type: seriesProps.type,
      data: [],
      filteredData: [],
      xAxisId: xAxis.id,
      yAxisId: yAxis.id
    };
  }
  createSeriesMetadata(_datum, seriesProps, xAxis, yAxis, seriesId, context) {
    const { intl, seriesIndexes } = context;
    const color = this.getColor(seriesProps.colorPalette, context.colorPalette);
    const seriesIndex = this.getOrCreateSeriesIndex(seriesId, seriesIndexes);
    const seriesName = this.getSeriesName(intl, seriesIndex, void 0);
    return {
      name: seriesName,
      xAxis,
      yAxis,
      actions: seriesProps.actions,
      valueFormatter: seriesProps.valueFormatter,
      valueUnit: seriesProps.valueUnit,
      colorPalette: color.colorPalette,
      valueAccessor: seriesProps.valueAccessor,
      colorIsFromSlot: color.colorIsFromSlot,
      valueType: void 0,
      valueAccessorLabel: seriesProps.valueAccessorLabel,
      valueMin: seriesProps.valueMin,
      valueMax: seriesProps.valueMax
    };
  }
  processImpl(datum, seriesProps, xAxis, yAxis, datapoint, context) {
    const { seriesMap, seriesToMetadata } = context;
    const seriesId = this.getSeriesId(datum, seriesProps);
    let currentSeries = seriesMap.get(seriesId);
    if (!currentSeries) {
      currentSeries = this.createSeries(seriesId, seriesProps, xAxis, yAxis);
      this.registerSeries(currentSeries, seriesId, context);
      const metadata = this.createSeriesMetadata(
        datum,
        seriesProps,
        xAxis,
        yAxis,
        seriesId,
        context
      );
      seriesToMetadata.set(currentSeries, {
        ...metadata,
        isVisible: context.filteredSeriesIds !== void 0 ? false : void 0
      });
    }
    const datapointMetadata = { isGap: true };
    if (isStringOrValidNumber(datapoint.value)) {
      const valueDomain = getValueScaleDomain(
        datapoint.value,
        seriesProps.id,
        this.seriesIdToValueScaleDomain
      );
      this.seriesIdToValueScaleDomain.set(seriesProps.id, valueDomain);
      datapointMetadata.isGap = false;
      const tempSeriesMetadata = safeGet(seriesToMetadata, currentSeries);
      tempSeriesMetadata.valueType ??= getLegendType(datapoint.value);
      const isDatapointVisible = tempSeriesMetadata.valueType === RANGE_LEGEND_TYPE || context.filteredSeriesIds === void 0 || context.filteredSeriesIds.has(`${seriesId}--${datapoint.value}`);
      datapointMetadata.isVisible = isDatapointVisible;
      tempSeriesMetadata.isVisible ||= isDatapointVisible;
    }
    this.registerDatapoint(
      datapoint,
      datapointMetadata,
      currentSeries,
      context
    );
    const datapointIsVisible = datapointMetadata.isVisible === void 0 || datapointMetadata.isVisible === true;
    if (datapointIsVisible) {
      context.axisWithSeries.add(xAxis.id);
      context.axisWithSeries.add(yAxis.id);
      if (!datapointMetadata.isGap) {
        this.addValue({
          series: currentSeries,
          datapoint
        });
      }
    }
    this.updateAxisDomains(
      datapoint,
      xAxis,
      yAxis,
      context.axisIdToScaleDomain,
      datapointIsVisible
    );
  }
  getColor(colorPalette, fallbackColorPalette) {
    const validColorPalette = getValidColorPalette(
      colorPalette,
      fallbackColorPalette ?? DEFAULT_COLOR_PALETTE_RECT
    );
    return {
      colorPalette: validColorPalette,
      colorIsFromSlot: validColorPalette === colorPalette
    };
  }
  getDatapoint(datum, seriesProps, xAxis, yAxis) {
    return getRectDatapoint(datum, seriesProps, xAxis, yAxis);
  }
}
export {
  RectSeriesStrategy
};
