// @dynatrace/strato-components/forms/forms-core/state/formfield-state-reducer.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var formfield_state_reducer_exports = {};
__export(formfield_state_reducer_exports, {
  FormFieldAction: () => FormFieldAction,
  _formFieldReducer: () => _formFieldReducer
});
module.exports = __toCommonJS(formfield_state_reducer_exports);
var FormFieldAction = /* @__PURE__ */ ((FormFieldAction2) => {
  FormFieldAction2["UPDATE"] = "update";
  return FormFieldAction2;
})(FormFieldAction || {});
function _formFieldReducer(state, action) {
  const { type, payload } = action;
  if (type === "update" /* UPDATE */) {
    return {
      controlId: payload.id || state.controlId,
      required: payload.required !== void 0 ? payload.required : state.required,
      disabled: payload.disabled !== void 0 ? payload.disabled : state.disabled,
      ariaDisabled: payload.ariaDisabled !== void 0 ? payload.ariaDisabled : state.ariaDisabled,
      readOnly: payload.readOnly !== void 0 ? payload.readOnly : state.readOnly
    };
  }
  return state;
}
