// @dynatrace/strato-components/forms/forms-core/components/label/Label.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Label_exports = {};
__export(Label_exports, {
  Label: () => Label
});
module.exports = __toCommonJS(Label_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_intl = require("react-intl");
var styles = __toESM(require("./Label.sty.js"));
var import_useBehavioralTrackingProps = require("../../../../core/hooks/useBehavioralTrackingProps.js");
var import_can_receive_default_text_format = require("../../../../core/utils/can-receive-default-text-format.js");
var import_merge_props = require("../../../../core/utils/merge-props.js");
var import_Text = require("../../../../typography/text/Text.js");
var import_useFormFieldContext = require("../../hooks/useFormFieldContext.js");
const COMPONENT_NAME = "Label";
const SimpleLabelText = (0, import_react.forwardRef)(
  (props, forwardedRef) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Text.Text,
    {
      ref: forwardedRef,
      as: "label",
      textStyle: "small-emphasized",
      ...props
    }
  )
);
const Label = /* @__PURE__ */ (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const {
      id,
      children,
      htmlFor,
      disabled: propDisabled,
      required: propRequired,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "aria-disabled": ariaDisabledProp,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
    const {
      controlId: formFieldContextId,
      required,
      disabled,
      ariaDisabled,
      readOnly
    } = (0, import_useFormFieldContext.useFormFieldContext)();
    const isStringChild = (0, import_can_receive_default_text_format.canReceiveDefaultTextFormat)(children);
    const Component = isStringChild ? SimpleLabelText : "label";
    const labelFor = htmlFor || formFieldContextId;
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      Component,
      {
        htmlFor: labelFor,
        id,
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        ref: forwardedRef,
        className: (0, import_clsx.default)(
          styles.label({
            disabled: disabled || propDisabled || ariaDisabled,
            // Only apply the required styles if we have string children, otherwise block level
            // elements in children will move the `*` required marker somewhere.
            required: isStringChild && (required ?? propRequired),
            readOnly
          }),
          consumerClassName
        ),
        style: consumerStyle,
        ...(0, import_merge_props.mergeProps)(restProps, behavioralTrackingProps),
        "data-dt-component": COMPONENT_NAME,
        children: [
          children,
          isStringChild && readOnly && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { textStyle: "small", className: (0, import_clsx.default)(styles.readOnlyLabel), children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_react_intl.FormattedMessage,
            {
              defaultMessage: "(read only)",
              id: "hIBEZw/eT6Nb+dz3",
              description: "Text appended to a label when the associated form field is read-only."
            }
          ) })
        ]
      }
    );
  }
);
Label.displayName = COMPONENT_NAME;
