// @dynatrace/strato-components/charts/histogram/components/legend/HistogramChartLegend.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var HistogramChartLegend_exports = {};
__export(HistogramChartLegend_exports, {
  HistogramChartLegend: () => HistogramChartLegend
});
module.exports = __toCommonJS(HistogramChartLegend_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
var import_useChartLayoutLoading = require("../../../core/components/chart-layout/hooks/useChartLayoutLoading.js");
var import_useChartLayoutOrientation = require("../../../core/components/chart-layout/hooks/useChartLayoutOrientation.js");
var import_orientation = require("../../../core/components/chart-layout/utils/orientation.js");
var import_explore_reducers = require("../../../core/components/interactions-state-helpers/reducers/explore-reducers.js");
var import_ChartLegend = require("../../../core/components/legend/categorical/ChartLegend.js");
var import_useLegendSeriesActions = require("../../../core/components/legend/categorical/hooks/useLegendSeriesActions.js");
var import_single_line_name = require("../../../core/utils/single-line-name.js");
var import_state_setters = require("../../hooks/store/state-setters.js");
var import_useHistogramData = require("../../hooks/useHistogramData.js");
var import_useHistogramTruncationMode = require("../../hooks/useHistogramTruncationMode.js");
var import_HistogramDataProvider = require("../../providers/HistogramDataProvider.js");
const createItemData = (series, seriesMap) => {
  const result = [];
  for (let index = 0; index < series.length; index++) {
    const singleSeries = series[index];
    const seriesMetadata = (0, import_HistogramDataProvider.safeGet)(seriesMap, singleSeries);
    result.push({
      id: seriesMetadata.id ?? `item-${index}`,
      color: seriesMetadata.color ?? import_colors.default.Text.Neutral.Default,
      name: (0, import_single_line_name.singleLineName)(singleSeries.name) ?? seriesMetadata.id
    });
  }
  return result;
};
const HistogramChartLegend = (0, import_react.memo)(() => {
  const { series, seriesMap } = (0, import_useHistogramData.useHistogramData)();
  const truncationMode = (0, import_useHistogramTruncationMode.useHistogramTruncationMode)();
  const setState = (0, import_state_setters.useSetState)();
  const { orientation: layoutOrientation } = (0, import_useChartLayoutOrientation.useChartLayoutOrientation)();
  const legendOrientation = (0, import_orientation.getLegendOrientation)(layoutOrientation);
  const isLoading = (0, import_useChartLayoutLoading.useChartLayoutLoading)();
  const legendItems = (0, import_react.useMemo)(
    () => createItemData(series, seriesMap),
    [series, seriesMap]
  );
  const updateHighlightedSeries = (item) => {
    setState((prev) => ({
      ...prev,
      highlightedSeriesId: item?.id
    }));
  };
  const updateVisibleSeries = (visibleSeries) => {
    const visibleSeriesSet = new Set(visibleSeries.map((s) => s.id));
    setState(
      (prev) => (0, import_explore_reducers.reduceExploreCancelAll)({
        ...prev,
        filteredSeriesIds: visibleSeriesSet
      })
    );
  };
  const legendActions = (0, import_useLegendSeriesActions.useLegendSeriesActions)(
    series,
    (s) => (0, import_HistogramDataProvider.safeGet)(seriesMap, s).id
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ChartLegend.ChartLegend,
    {
      items: legendItems,
      orientation: legendOrientation,
      interactionsDisabled: isLoading,
      onHover: updateHighlightedSeries,
      onClick: updateVisibleSeries,
      truncationMode,
      onSeriesAction: legendActions
    }
  );
});
HistogramChartLegend["displayName"] = "HistogramChartLegend";
