// @dynatrace/strato-components/esm/forms/calendar/calendar-header/CalendarHeader.js
import { jsx, jsxs } from "react/jsx-runtime";
import { forwardRef } from "react";
import { useIntl } from "react-intl";
import {
  ChevronLeftSmallIcon,
  ChevronRightSmallIcon
} from "@dynatrace/strato-icons";
import { Button } from "../../../buttons/button/Button.js";
import { Flex } from "../../../layouts/flex/Flex.js";
import { useFormFieldContext } from "../../forms-core/hooks/useFormFieldContext.js";
import { SelectDateLabel } from "../calendar-header-label/SelectDateLabel.js";
import { SelectMonthLabel } from "../calendar-header-label/SelectMonthLabel.js";
import { SelectYearLabel } from "../calendar-header-label/SelectYearLabel.js";
import * as itemStyles from "../components/CalendarItem.sty.js";
import { useCalendarContext } from "../hooks/useCalendarContext.js";
import { NUMBER_OF_YEARS } from "../types/calendarConfiguration.js";
function getViewSpecificProps(view) {
  switch (view) {
    case "day": {
      const nextView = "month";
      return {
        label: /* @__PURE__ */ jsx(SelectDateLabel, {}),
        nextView,
        headerDataDtProperties: `value:${nextView}`,
        payload: { amount: 1, unit: nextView }
      };
    }
    case "month": {
      const nextView = "year";
      return {
        label: /* @__PURE__ */ jsx(SelectMonthLabel, {}),
        nextView,
        headerDataDtProperties: `value:${nextView}`,
        payload: { amount: 1, unit: nextView }
      };
    }
    case "year": {
      const nextView = "day";
      return {
        label: /* @__PURE__ */ jsx(SelectYearLabel, {}),
        nextView,
        headerDataDtProperties: `value:${nextView}`,
        payload: { amount: NUMBER_OF_YEARS }
      };
    }
  }
}
const CalendarHeader = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const { labelId, view } = props;
  const { dispatch } = useCalendarContext();
  const { disabled, ariaDisabled } = useFormFieldContext();
  const intl = useIntl();
  const { label, nextView, payload, headerDataDtProperties } = getViewSpecificProps(view);
  return /* @__PURE__ */ jsxs(Flex, { justifyContent: "space-between", marginBottom: 16, children: [
    /* @__PURE__ */ jsx(
      Button,
      {
        disabled,
        "aria-disabled": ariaDisabled,
        "aria-label": intl.formatMessage({
          defaultMessage: "Go to previous page.",
          id: "VoqQWp9yEM0QM8eF",
          description: "Aria label to add to the `previous` button"
        }),
        onClick: () => dispatch({
          type: "navigateCurrentView",
          payload: { ...payload, amount: payload.amount * -1 }
        }),
        "data-dt-name": "PreviousPage",
        children: /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(ChevronLeftSmallIcon, {}) })
      }
    ),
    /* @__PURE__ */ jsx(
      Button,
      {
        className: itemStyles.headerButton,
        disabled,
        "aria-disabled": ariaDisabled,
        "aria-live": "polite",
        id: labelId,
        onClick: () => dispatch({
          type: "currentViewChange",
          payload: nextView
        }),
        ref: forwardedRef,
        "data-dt-name": "ChangeView",
        "data-dt-properties": headerDataDtProperties,
        children: label
      }
    ),
    /* @__PURE__ */ jsx(
      Button,
      {
        disabled,
        "aria-disabled": ariaDisabled,
        "aria-label": intl.formatMessage({
          defaultMessage: "Go to next page.",
          id: "KAAzM9PVtiJno9ks",
          description: "Aria label to add to the `next` button"
        }),
        onClick: () => dispatch({
          type: "navigateCurrentView",
          payload
        }),
        "data-dt-name": "NextPage",
        children: /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(ChevronRightSmallIcon, {}) })
      }
    )
  ] });
});
CalendarHeader.displayName = "CalendarHeader";
export {
  CalendarHeader
};
