// @dynatrace/strato-components/esm/charts/core/components/single-value-renderer/constants.js
import Colors from "@dynatrace/strato-design-tokens/colors";
const FONT_HEIGHT_RATIO = 0.7;
const FONT_WIDTH_RATIO = 0.95;
const VALUE_FONT_CSS_TOKEN = "--dt-charts-single-value-font-size";
const LABEL_FONT_CSS_TOKEN = "--dt-charts-single-value-label-font-size";
const TREND_FONT_CSS_TOKEN = "--dt-charts-single-value-label-font-size";
const DEFAULT_MAX_VALUE_FONT_SIZE = 240;
const DEFAULT_SINGLE_VALUE_COLOR = Colors.Text.Neutral.Default;
const GLOBAL_VALUE_FONT_CSS_TOKEN = "--dt-charts-single-value-global-value-font-size";
const GLOBAL_LABEL_FONT_CSS_TOKEN = "--dt-charts-single-value-global-label-font-size";
const GLOBAL_TREND_FONT_CSS_TOKEN = "--dt-charts-single-value-global-trend-font-size";
const MAX_VALUE_FONT_SIZE = `var(${GLOBAL_VALUE_FONT_CSS_TOKEN}, ${DEFAULT_MAX_VALUE_FONT_SIZE})`;
const VALUE_FONT_SIZE = `min(var(${VALUE_FONT_CSS_TOKEN}, ${MAX_VALUE_FONT_SIZE}), ${MAX_VALUE_FONT_SIZE})`;
const MAX_LABEL_FONT_SIZE = `min(var(${GLOBAL_LABEL_FONT_CSS_TOKEN}, ${VALUE_FONT_SIZE}), ${VALUE_FONT_SIZE})`;
const LABEL_FONT_SIZE = `min(var(${LABEL_FONT_CSS_TOKEN}, ${MAX_LABEL_FONT_SIZE}), ${MAX_LABEL_FONT_SIZE})`;
const MAX_TREND_FONT_SIZE = `min(var(${GLOBAL_TREND_FONT_CSS_TOKEN}, ${VALUE_FONT_SIZE}), ${VALUE_FONT_SIZE})`;
const TREND_FONT_SIZE = `min(var(${LABEL_FONT_CSS_TOKEN}, ${MAX_TREND_FONT_SIZE}), ${MAX_TREND_FONT_SIZE})`;
const UPWARD_ICON_DEFAULT_COLOR = Colors.Icon.Success.Default;
const NEUTRAL_ICON_DEFAULT_COLOR = Colors.Icon.Primary.Default;
const DOWNWARD_ICON_DEFAULT_COLOR = Colors.Icon.Critical.Default;
const UPWARD_VALUE_DEFAULT_COLOR = Colors.Text.Success.Default;
const NEUTRAL_VALUE_DEFAULT_COLOR = Colors.Text.Primary.Default;
const DOWNWARD_VALUE_DEFAULT_COLOR = Colors.Text.Critical.Default;
const TREND_LABEL_DEFAULT_COLOR = Colors.Text.Neutral.Default;
const MINIMUM_HEIGHT_FOR_TOOLBAR = 40;
const MINIMUM_HEIGHT_FOR_TOOLBAR_WITH_PADDING = 48;
export {
  DEFAULT_MAX_VALUE_FONT_SIZE,
  DEFAULT_SINGLE_VALUE_COLOR,
  DOWNWARD_ICON_DEFAULT_COLOR,
  DOWNWARD_VALUE_DEFAULT_COLOR,
  FONT_HEIGHT_RATIO,
  FONT_WIDTH_RATIO,
  GLOBAL_LABEL_FONT_CSS_TOKEN,
  GLOBAL_TREND_FONT_CSS_TOKEN,
  GLOBAL_VALUE_FONT_CSS_TOKEN,
  LABEL_FONT_CSS_TOKEN,
  LABEL_FONT_SIZE,
  MINIMUM_HEIGHT_FOR_TOOLBAR,
  MINIMUM_HEIGHT_FOR_TOOLBAR_WITH_PADDING,
  NEUTRAL_ICON_DEFAULT_COLOR,
  NEUTRAL_VALUE_DEFAULT_COLOR,
  TREND_FONT_CSS_TOKEN,
  TREND_FONT_SIZE,
  TREND_LABEL_DEFAULT_COLOR,
  UPWARD_ICON_DEFAULT_COLOR,
  UPWARD_VALUE_DEFAULT_COLOR,
  VALUE_FONT_CSS_TOKEN,
  VALUE_FONT_SIZE
};
