// @dynatrace/strato-components/esm/charts/core/components/single-value-renderer/SingleValueRenderer.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import { isEmpty, isUndefined } from "lodash-es";
import {
  useRef
} from "react";
import { ContentGrid } from "./components/ContentGrid.js";
import { Label } from "./components/Label.js";
import {
  SingleValueToolbar
} from "./components/SingleValueToolbar.js";
import { Trend } from "./components/trend/Trend.js";
import { Value } from "./components/Value.js";
import {
  SingleValueContext
} from "./contexts/single-value.context.js";
import { FontSizeProvider } from "./providers/font-size.provider.js";
import {
  chartStateHandler,
  singleValueContainer,
  singleValueInnerContainer,
  sparklineArea,
  trendArea
} from "./SingleValueRenderer.sty.js";
import { buildValueContentGridRows } from "./utils/build-value-content-grid-rows.js";
import { formatLabel } from "./utils/format-label.js";
import { formatValue } from "./utils/format-value.js";
import { getCalculatedColor } from "./utils/get-calculated-color.js";
import { getContentAndBackgroundColor } from "./utils/get-content-and-background-color.js";
import { useContainerSize } from "../../../../core/hooks/useContainerSize.js";
import { defaultChartSize } from "../../../../core/styles/defaultChartValues.sty.js";
import { ChartStateHandler } from "../chart-states/ChartStateHandler.js";
import { LoaderOverlay } from "../loader-overlay/LoaderOverlay.js";
import { SparklineDisplay } from "../sparkline-display/SparklineDisplay.js";
const SingleValueRenderer = (props) => {
  const {
    data,
    formatter,
    unit,
    prefixIcon,
    alignment,
    label,
    color,
    colorRules = [],
    emptyState,
    errorState,
    loading = false,
    sparklineProps,
    sparklineChildren,
    trendProps,
    applyThresholdBackground,
    propsFromGrid,
    toolbar,
    dataItem
  } = props;
  const containerRef = useRef(null);
  const { height: containerHeight } = useContainerSize(containerRef);
  const { color: calculatedColor, ruleApplied } = getCalculatedColor(
    data,
    colorRules,
    color,
    dataItem
  );
  const applyBackgroundColor = applyThresholdBackground && ruleApplied;
  const { contentColor, backgroundColor } = getContentAndBackgroundColor(
    calculatedColor,
    applyBackgroundColor
  );
  const { value, unit: formatterUnit } = formatValue(data, unit, formatter);
  const formattedLabel = formatLabel(label);
  const context = {
    value,
    unit: formatterUnit ?? "",
    label: formattedLabel ?? "",
    prefixIcon,
    alignment
  };
  const { innerGrid, outerGrid } = buildValueContentGridRows(
    !isUndefined(label),
    !isUndefined(data),
    !isUndefined(trendProps),
    !isUndefined(sparklineProps)
  );
  const getSparklineColor = () => applyBackgroundColor ? contentColor : sparklineProps?.color ?? contentColor;
  return /* @__PURE__ */ jsx(
    "div",
    {
      className: singleValueContainer,
      style: { backgroundColor },
      ref: containerRef,
      children: /* @__PURE__ */ jsx(
        FontSizeProvider,
        {
          text: context.prefixIcon ? `00${context.value}` : `${context.value}`,
          suffix: context.unit,
          children: /* @__PURE__ */ jsxs(SingleValueContext.Provider, { value: context, children: [
            /* @__PURE__ */ jsxs(ContentGrid, { color: contentColor, gridTemplate: outerGrid, children: [
              label !== void 0 ? /* @__PURE__ */ jsx(Label, { color: applyBackgroundColor ? contentColor : void 0 }) : null,
              /* @__PURE__ */ jsx("div", { className: clsx(singleValueInnerContainer), children: /* @__PURE__ */ jsxs(
                ChartStateHandler,
                {
                  empty: isEmpty(value),
                  height: "100cqh",
                  emptyState,
                  errorState,
                  isLoading: loading,
                  style: {
                    gridTemplate: innerGrid
                  },
                  className: clsx(defaultChartSize, chartStateHandler),
                  children: [
                    /* @__PURE__ */ jsx(Value, {}),
                    !isUndefined(trendProps) ? /* @__PURE__ */ jsx("div", { className: trendArea, children: /* @__PURE__ */ jsx(
                      Trend,
                      {
                        ...trendProps,
                        contrastColor: applyBackgroundColor ? contentColor : void 0
                      }
                    ) }) : null,
                    !isUndefined(sparklineProps) ? /* @__PURE__ */ jsx("div", { className: sparklineArea, children: /* @__PURE__ */ jsx(
                      SparklineDisplay,
                      {
                        ...sparklineProps,
                        color: getSparklineColor(),
                        labelsAlignment: "right",
                        labelsFixedWidth: void 0,
                        ticksColor: applyBackgroundColor ? contentColor : void 0,
                        style: {
                          display: "flex",
                          flexDirection: "column",
                          justifyContent: "flex-end"
                        },
                        children: sparklineChildren
                      }
                    ) }) : null,
                    loading && /* @__PURE__ */ jsx(LoaderOverlay, {})
                  ]
                }
              ) })
            ] }),
            /* @__PURE__ */ jsx(
              SingleValueToolbar,
              {
                ...propsFromGrid?.toolbar ? propsFromGrid?.toolbar : toolbar,
                parentHeight: containerHeight,
                intentsAsyncTemplate: propsFromGrid?.intentsAsyncTemplate
              }
            )
          ] })
        }
      )
    }
  );
};
export {
  SingleValueRenderer
};
