// @dynatrace/strato-components/editors/base-code-editor/highlight/highlight.extension.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var highlight_extension_exports = {};
__export(highlight_extension_exports, {
  highlightExtension: () => highlightExtension
});
module.exports = __toCommonJS(highlight_extension_exports);
var import_state = require("@codemirror/state");
var import_view = require("@codemirror/view");
const highlight = import_view.Decoration.mark({
  attributes: {
    style: "animation: none"
  },
  class: "cm-highlight"
});
function highlightUpdatedLines(view, previousValue) {
  const builder = new import_state.RangeSetBuilder();
  const previousSplittedValue = previousValue.split("\n");
  if (previousValue) {
    for (const { from, to } of view.visibleRanges) {
      for (let pos = from; pos <= to; ) {
        const line = view.state.doc.lineAt(pos);
        if (previousSplittedValue[line.number - 1] !== line.text) {
          builder.add(line.from, line.to, highlight);
        }
        pos = line.to + 1;
      }
    }
  }
  return builder.finish();
}
const highlightExtension = (previousValue) => {
  const highlightLines = import_view.ViewPlugin.fromClass(
    class {
      decorations;
      constructor(view) {
        this.decorations = highlightUpdatedLines(view, previousValue);
        requestAnimationFrame(() => {
          view?.dom.querySelectorAll(".cm-highlight").forEach((n) => {
            n.style.animation = "";
          });
        });
      }
      /** Run update if user changed input/focus */
      update(update) {
        if (update.docChanged || update.focusChanged) {
          this.decorations = new import_state.RangeSetBuilder().finish();
        }
      }
    },
    {
      decorations: (v) => v.decorations
    }
  );
  return import_state.Prec.default(highlightLines);
};
