// @dynatrace/strato-components/forms/calendar/CalendarRoot.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CalendarRoot_exports = {};
__export(CalendarRoot_exports, {
  CalendarRoot: () => CalendarRoot
});
module.exports = __toCommonJS(CalendarRoot_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_CalendarHeader = require("./calendar-header/CalendarHeader.js");
var import_calendarMessages = require("./calendarMessages.js");
var import_VisuallyHidden = require("../../core/components/visually-hidden/VisuallyHidden.js");
var import_useTimeframeStore = require("../../core/time/timeframeStore/useTimeframeStore.js");
var import_validation = require("../../core/time/validation.js");
var import_useFormFieldMessagesContext = require("../forms-core/hooks/useFormFieldMessagesContext.js");
var import_useFormFieldValidation = require("../forms-core/hooks/useFormFieldValidation.js");
var import_formfield_messages_state_reducer = require("../forms-core/state/formfield-messages-state-reducer.js");
var styles = __toESM(require("./Calendar.sty.js"));
var import_CalendarContext = require("./contexts/CalendarContext.js");
var import_useCalendar = require("./hooks/useCalendar.js");
var import_useCalendarNavigation = require("./hooks/useCalendarNavigation.js");
var import_SelectDayView = require("./views/SelectDayView.js");
var import_SelectMonthView = require("./views/SelectMonthView.js");
var import_SelectYearView = require("./views/SelectYearView.js");
var import_helpers = require("../../core/time/helpers.js");
const COMPONENT_NAME = "Calendar";
function renderView(view) {
  switch (view) {
    case "month":
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_SelectMonthView.SelectMonthView, {});
    case "year":
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_SelectYearView.SelectYearView, {});
    default:
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_SelectDayView.SelectDayView, {});
  }
}
function CalendarComponent(props, forwardedRef) {
  const {
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    id,
    disabled,
    "aria-disabled": ariaDisabled,
    required,
    name,
    readOnly = false,
    // eslint-disable-next-line @typescript-eslint/prefer-ts-expect-error, @typescript-eslint/ban-ts-comment
    // @ts-ignore Strict mode transition - needs to be investigated
    type = "single",
    "aria-label": ariaLabelProp,
    "aria-labelledby": ariaLabelledByProp,
    "aria-describedby": ariaDescribedByProp,
    "aria-details": ariaDetailsProp,
    ...remainingProps
  } = props;
  const calendarValue = (0, import_useTimeframeStore.useCalendarStore)((state2) => state2.calendarValue);
  const min = (0, import_useTimeframeStore.useCalendarStore)((state2) => state2.min);
  const max = (0, import_useTimeframeStore.useCalendarStore)((state2) => state2.max);
  const error = (0, import_useTimeframeStore.useCalendarStore)((state2) => state2.calendarError);
  const precision = (0, import_useTimeframeStore.useCalendarStore)((state2) => state2.precision);
  const calendarRef = (0, import_react.useRef)(null);
  const ref = (0, import_react.useRef)(null);
  const refTo = (0, import_react.useRef)(null);
  const { uniqueId, state, dispatch } = (0, import_useCalendar.useCalendar)({
    type
  });
  (0, import_useCalendarNavigation.useCalendarNavigation)({
    calendarRef,
    view: state.currentView,
    dispatch
  });
  const { dispatch: formMessageDispatch } = (0, import_useFormFieldMessagesContext.useFormFieldMessagesContext)();
  const refValidate = (0, import_react.useCallback)(() => {
    formMessageDispatch?.({
      type: import_formfield_messages_state_reducer.FormFieldMessagesAction.SETISTOUCHED,
      payload: { isTouched: true }
    });
    (0, import_validation.validate)(calendarValue, min, max, {
      formMessageDispatch,
      formerError: error,
      precision,
      showError: true,
      checkOrder: false,
      ref,
      errorMessages: import_calendarMessages.CALENDAR_ERROR_MESSAGES
    });
  }, [min, max, formMessageDispatch, error, precision, ref, calendarValue]);
  (0, import_react.useImperativeHandle)(
    forwardedRef,
    () => {
      if (type === "single") {
        return {
          element: calendarRef.current,
          inputRef: ref.current,
          validate: refValidate
        };
      }
      return {
        element: calendarRef.current,
        inputRef: {
          from: ref?.current,
          to: refTo?.current
        },
        validate: refValidate
      };
    },
    [refValidate, type]
  );
  const {
    ariaProps,
    onChange: validationOnChange,
    onBlur,
    onInvalid
  } = (0, import_useFormFieldValidation.useFormFieldValidation)(
    ref,
    {},
    {
      "aria-label": ariaLabelProp,
      "aria-labelledby": ariaLabelledByProp,
      "aria-describedby": ariaDescribedByProp,
      "aria-details": ariaDetailsProp
    }
  );
  const labelRef = (0, import_react.useRef)(`calendar-body-label-${uniqueId}`);
  const disabledOrAriaDisabled = disabled || ariaDisabled || false;
  const getNativeProps = (suffix = "") => ({
    readOnly,
    required,
    disabled,
    "aria-disabled": ariaDisabled,
    name: name && `${name}${suffix}`,
    id
  });
  const calenderContextValue = (0, import_react.useMemo)(
    () => ({ state, dispatch, readOnly }),
    [state, dispatch, readOnly]
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_CalendarContext.CalendarContext.Provider, { value: calenderContextValue, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      "div",
      {
        className: (0, import_clsx.default)(styles.calendar({ readOnly }), consumerClassName, {
          [styles.disabled]: disabledOrAriaDisabled
        }),
        "data-testid": dataTestId,
        ref: calendarRef,
        style: consumerStyle,
        ...remainingProps,
        "data-dt-component": COMPONENT_NAME,
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_CalendarHeader.CalendarHeader, { view: state.currentView, labelId: labelRef.current }),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { "aria-labelledby": labelRef.current, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "table",
            {
              role: "grid",
              className: styles.body({
                isReadOnly: !!readOnly
              }),
              children: renderView(state.currentView)
            }
          ) })
        ]
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_VisuallyHidden.VisuallyHidden, { children: (0, import_helpers.isTimeValue)(calendarValue) ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "input",
      {
        onChange: validationOnChange,
        onBlur,
        onInvalid,
        ...ariaProps,
        ...getNativeProps(),
        ref,
        value: calendarValue?.absoluteDate ?? "",
        tabIndex: -1
      }
    ) : /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        "input",
        {
          onChange: validationOnChange,
          onBlur,
          onInvalid,
          ...ariaProps,
          ...getNativeProps("-from"),
          ref,
          value: (0, import_helpers.isFullTimeframe)(calendarValue) && calendarValue?.from?.value ? calendarValue.from.value : "",
          tabIndex: -1
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        "input",
        {
          ...getNativeProps("-to"),
          ref: refTo,
          value: calendarValue?.to?.value ?? "",
          id: `${id}-to`,
          type: "hidden"
        }
      )
    ] }) })
  ] });
}
CalendarComponent["displayName"] = COMPONENT_NAME;
const CalendarRoot = (0, import_react.forwardRef)(CalendarComponent);
