// @dynatrace/strato-components/forms/calendar/Calendar.sty.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Calendar_css_exports = {};
__export(Calendar_css_exports, {
  body: () => body,
  calendar: () => calendar,
  clearBackground: () => clearBackground,
  dayRow: () => dayRow,
  disabled: () => disabled,
  field: () => field,
  monthRow: () => monthRow,
  rangeField: () => rangeField,
  row: () => row,
  weekdays: () => weekdays,
  yearRow: () => yearRow
});
module.exports = __toCommonJS(Calendar_css_exports);
var import_Calendar_css_ts_vanilla = require("./Calendar.css");
var import_createRuntimeFn = require("@vanilla-extract/recipes/createRuntimeFn");
var body = (0, import_createRuntimeFn.createRuntimeFn)({ defaultClassName: "_xm8mdk3-3-6-0", variantClassNames: { isReadOnly: { true: "_xm8mdk4-3-6-0" } }, defaultVariants: {}, compoundVariants: [] });
var calendar = (0, import_createRuntimeFn.createRuntimeFn)({ defaultClassName: "_xm8mdk0-3-6-0", variantClassNames: { readOnly: { true: "_xm8mdk1-3-6-0" } }, defaultVariants: {}, compoundVariants: [] });
var clearBackground = "_xm8mdkm-3-6-0";
var dayRow = "_xm8mdk6-3-6-0";
var disabled = "_xm8mdk2-3-6-0";
var field = (0, import_createRuntimeFn.createRuntimeFn)({ defaultClassName: "_xm8mdk9-3-6-0", variantClassNames: { color: { primary: "_xm8mdka-3-6-0", neutral: "_xm8mdkb-3-6-0", critical: "_xm8mdkc-3-6-0" }, isIncompleteSelection: { true: "_xm8mdkd-3-6-0" }, isDisabled: { true: "_xm8mdke-3-6-0" } }, defaultVariants: { isDisabled: false, color: "neutral" }, compoundVariants: [[{ color: "primary", isDisabled: true }, "_xm8mdkf-3-6-0"], [{ color: "neutral", isDisabled: true }, "_xm8mdkg-3-6-0"], [{ color: "critical", isDisabled: true }, "_xm8mdkh-3-6-0"]] });
var monthRow = "_xm8mdk7-3-6-0";
var rangeField = (0, import_createRuntimeFn.createRuntimeFn)({ defaultClassName: "_xm8mdki-3-6-0", variantClassNames: { rangeKind: { first: "_xm8mdkj-3-6-0", last: "_xm8mdkk-3-6-0", inbetween: "_xm8mdkl-3-6-0" } }, defaultVariants: {}, compoundVariants: [] });
var row = "_xm8mdk5-3-6-0";
var weekdays = "_xm8mdkn-3-6-0";
var yearRow = "_xm8mdk8-3-6-0";
