// @dynatrace/strato-components/forms/calendar/Calendar.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Calendar_exports = {};
__export(Calendar_exports, {
  Calendar: () => Calendar
});
module.exports = __toCommonJS(Calendar_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_CalendarRoot = require("./CalendarRoot.js");
var import_helpers = require("./utils/helpers.js");
var import_useCallbackRef = require("../../core/hooks/useCallbackRef.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_helpers2 = require("../../core/time/helpers.js");
var import_calendar = require("../../core/time/timeframeStore/actions/calendar.js");
var import_createTimeframeStore = require("../../core/time/timeframeStore/createTimeframeStore.js");
var import_timeframeStoreProvider = require("../../core/time/timeframeStore/timeframeStoreProvider.js");
var import_useMinMax = require("../../core/time/useMinMax.js");
var import_validation = require("../../core/time/validation.js");
var import_useFormFieldMessagesContext = require("../forms-core/hooks/useFormFieldMessagesContext.js");
var import_useSyncedFormControlProps = require("../forms-core/hooks/useSyncedFormControlProps.js");
var import_calendarMessages = require("./calendarMessages.js");
function useSyncPropsAndState(store, memoedValue, isControlled, type, precision, min, max, formFieldMessageDispatch, calendarRef) {
  const valueMode = type === "single" ? "single" : "timeframeSelector";
  (0, import_react.useEffect)(
    () => {
      isControlled && store.setState((state) => {
        const error = (0, import_validation.validate)(memoedValue, state.min, state.max, {
          formMessageDispatch: formFieldMessageDispatch,
          formerError: state.calendarError,
          precision: state.precision,
          showError: false,
          checkOrder: true,
          ref: calendarRef,
          valueMode,
          errorMessages: import_calendarMessages.CALENDAR_ERROR_MESSAGES
        });
        const timeOffset = (0, import_helpers.getTimeOffsetFromInitialValue)(
          memoedValue,
          type,
          min,
          max,
          precision
        );
        const valueWithPrecision = (0, import_helpers.convertToCalendarStructure)(
          type,
          memoedValue,
          min,
          max,
          precision
        );
        if (type === "single") {
          return {
            calendarValue: valueWithPrecision,
            calendarError: error,
            timeOffset
          };
        }
        const value = (0, import_helpers2.isFullTimeframe)(valueWithPrecision) ? valueWithPrecision : null;
        return {
          calendarValue: value,
          calendarError: error,
          timeOffset
        };
      });
    },
    // calendarRef is a stable ref — no need to include it in deps
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [
      formFieldMessageDispatch,
      isControlled,
      memoedValue,
      store,
      min,
      max,
      precision,
      type
    ]
  );
  (0, import_react.useEffect)(() => {
    store.setState((state) => {
      const value = (0, import_helpers.convertToCalendarStructure)(
        type,
        state.calendarValue,
        min,
        max,
        precision
      );
      return {
        ...state,
        precision,
        calendarValue: value
      };
    });
  }, [precision, store, min, max, type]);
}
function RefAwareCalendar(props, forwardedRef) {
  const {
    value: propValue,
    defaultValue: propDefaultValue,
    onChange,
    min: propMin,
    max: propMax,
    disabled: propDisabled,
    required: propRequired,
    readOnly: propReadOnly,
    id: propId,
    "aria-disabled": propAriaDisabled,
    precision = "minutes",
    type = "single",
    ...remainingProps
  } = props;
  const [isControlled] = (0, import_react.useState)(propValue !== void 0);
  const onChangeProp = (0, import_useCallbackRef.useCallbackRef)(onChange);
  const { min, max } = (0, import_useMinMax.useMinMax)(propMin, propMax);
  const memoedValue = (0, import_react.useMemo)(() => {
    return (0, import_helpers.convertToCalendarStructure)(type, propValue, min, max, precision);
  }, [precision, propValue, min, max, type]);
  const memoedDefaultValue = (0, import_react.useMemo)(
    () => (0, import_helpers.convertToCalendarStructure)(type, propDefaultValue, min, max, precision),
    /** We're ignoring updates to defaultValue as it only considers the initial state. */
    // eslint-disable-next-line react-hooks/exhaustive-deps
    []
  );
  const calendarRef = (0, import_react.useRef)(null);
  const { dispatch: formFieldMessageDispatch } = (0, import_useFormFieldMessagesContext.useFormFieldMessagesContext)();
  const { disabled, required, id, ariaDisabled, readOnly } = (0, import_useSyncedFormControlProps.useSyncedFormControlProps)({
    propDisabled,
    propRequired,
    propId,
    propAriaDisabled,
    propReadOnly
  });
  const [store] = (0, import_react.useState)(() => {
    const value = isControlled ? memoedValue : memoedDefaultValue;
    const error = (0, import_validation.validate)(value, min, max, {
      formMessageDispatch: formFieldMessageDispatch,
      precision,
      showError: false,
      checkOrder: true,
      ref: calendarRef,
      valueMode: type === "single" ? "single" : "timeframeSelector",
      required,
      errorMessages: import_calendarMessages.CALENDAR_ERROR_MESSAGES
    });
    const timeOffset = (0, import_helpers.getTimeOffsetFromInitialValue)(
      value,
      type,
      min,
      max,
      precision
    );
    const initialState = {
      calendarValue: value,
      min,
      max,
      precision,
      isFullDay: false,
      calendarRef,
      required,
      calendarError: error,
      timeOffset
    };
    const actions = (0, import_calendar.getCalendarActions)(
      isControlled,
      type,
      formFieldMessageDispatch,
      onChangeProp
    );
    return (0, import_createTimeframeStore.createTimeframeStore)(initialState, actions);
  });
  useSyncPropsAndState(
    store,
    memoedValue,
    isControlled,
    type,
    precision,
    min,
    max,
    formFieldMessageDispatch,
    calendarRef
  );
  (0, import_useMinMax.useMinMaxStoreSync)(
    store,
    min,
    max,
    (0, import_react.useCallback)(
      (state, min2, max2) => ({
        calendarError: (0, import_validation.validate)(state.calendarValue, min2, max2, {
          formMessageDispatch: formFieldMessageDispatch,
          precision,
          showError: false,
          checkOrder: true,
          ref: calendarRef,
          valueMode: type === "single" ? "single" : "timeframeSelector",
          required,
          errorMessages: import_calendarMessages.CALENDAR_ERROR_MESSAGES
        })
      }),
      [formFieldMessageDispatch]
    )
  );
  const mergedRef = (0, import_useMergeRefs.useMergeRefs)([calendarRef, forwardedRef]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_timeframeStoreProvider.CalendarStoreProvider, { store, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_CalendarRoot.CalendarRoot,
    {
      ...remainingProps,
      id,
      ref: mergedRef,
      "aria-disabled": ariaDisabled,
      disabled,
      required,
      readOnly,
      type
    }
  ) });
}
const Calendar = (0, import_react.forwardRef)(RefAwareCalendar);
