// @dynatrace/strato-components/forms/calendar/calendarMessages.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var calendarMessages_exports = {};
__export(calendarMessages_exports, {
  CALENDAR_ERROR_MESSAGES: () => CALENDAR_ERROR_MESSAGES
});
module.exports = __toCommonJS(calendarMessages_exports);
var import_react_intl = require("react-intl");
const CALENDAR_ERROR_MESSAGES = (0, import_react_intl.defineMessages)({
  minBoundary: {
    id: "9XjoWQoeIKxStpgZ",
    defaultMessage: "Please select a {term} later than {min}.",
    description: "Error message for a date or a timeframe earlier than the required minimum."
  },
  maxBoundary: {
    id: "0Y8wN2jk+BAa2NoC",
    defaultMessage: "Please select a {term} earlier than {max}.",
    description: "Error message for a date or a timeframe later then the expected maximum."
  },
  minMaxBoundary: {
    id: "LNWPcycjuDPimFKd",
    defaultMessage: "Please select a {term} within the range of {min} and {max}.",
    description: "Error message for a date or a timeframe outside the selectable range."
  }
});
