// @dynatrace/strato-components/esm/charts/pie/components/InnerWrapper/components/InnerDonut/InnerDonut.js
import { jsx } from "react/jsx-runtime";
import { isFunction, round } from "lodash-es";
import { useMemo } from "react";
import { container } from "./InnerDonut.sty.js";
import { Flex } from "../../../../../../layouts/flex/Flex.js";
import { useChartInnerElement } from "../../../../context/InnerElementContext.js";
import { useVisibleSlicesState } from "../../../../hooks/store/selectors.js";
import { buildCumulativeValues } from "../../../../utils/build-cumulative-value.js";
const InnerDonut = ({ donutInner, visibleIds }) => {
  const { innerElement } = useChartInnerElement();
  const visibleSlices = useVisibleSlicesState();
  const { absoluteValue, relativeValue } = useMemo(
    () => buildCumulativeValues(visibleSlices, visibleIds),
    [visibleIds, visibleSlices]
  );
  if (!innerElement) {
    return null;
  }
  const containerStyle = {
    width: `${donutInner}px`,
    height: `${donutInner}px`
  };
  return /* @__PURE__ */ jsx(
    Flex,
    {
      justifyContent: "center",
      alignItems: "center",
      className: container,
      style: containerStyle,
      children: isFunction(innerElement) ? innerElement({
        absoluteValue,
        relativeValue: round(relativeValue, 2),
        dimensions: { width: donutInner, height: donutInner }
      }) : innerElement
    }
  );
};
export {
  InnerDonut
};
