// @dynatrace/strato-components/esm/core/time/timeframeStore/timeframeStoreProvider.js
import { jsx } from "react/jsx-runtime";
import { createContext } from "react";
const CalendarStoreCtx = createContext(null);
const DateTimePickerStoreCtx = createContext(null);
const DateTimeRangePickerStoreCtx = createContext(null);
const TimeframeSelectorStoreCtx = createContext(null);
function CalendarStoreProvider({
  store,
  children
}) {
  return /* @__PURE__ */ jsx(CalendarStoreCtx.Provider, { value: store, children });
}
function DateTimePickerStoreProvider({
  store,
  children
}) {
  const calStore = store;
  return /* @__PURE__ */ jsx(CalendarStoreCtx.Provider, { value: calStore, children: /* @__PURE__ */ jsx(DateTimePickerStoreCtx.Provider, { value: store, children }) });
}
function DateTimeRangePickerStoreProvider({
  store,
  children
}) {
  const calStore = store;
  const dtpStore = store;
  return /* @__PURE__ */ jsx(CalendarStoreCtx.Provider, { value: calStore, children: /* @__PURE__ */ jsx(DateTimePickerStoreCtx.Provider, { value: dtpStore, children: /* @__PURE__ */ jsx(DateTimeRangePickerStoreCtx.Provider, { value: store, children }) }) });
}
function TimeframeSelectorStoreProvider({
  store,
  children
}) {
  const calStore = store;
  const dtpStore = store;
  const dtrpStore = store;
  return /* @__PURE__ */ jsx(CalendarStoreCtx.Provider, { value: calStore, children: /* @__PURE__ */ jsx(DateTimePickerStoreCtx.Provider, { value: dtpStore, children: /* @__PURE__ */ jsx(DateTimeRangePickerStoreCtx.Provider, { value: dtrpStore, children: /* @__PURE__ */ jsx(TimeframeSelectorStoreCtx.Provider, { value: store, children }) }) }) });
}
export {
  CalendarStoreCtx,
  CalendarStoreProvider,
  DateTimePickerStoreCtx,
  DateTimePickerStoreProvider,
  DateTimeRangePickerStoreCtx,
  DateTimeRangePickerStoreProvider,
  TimeframeSelectorStoreCtx,
  TimeframeSelectorStoreProvider
};
