// @dynatrace/strato-components/esm/charts/meter-bar/single-meter-bar/utils/build-meter-bar-chart-config.js
import { applyThresholdDefaults } from "./apply-threshold-defaults.js";
import { isValidColorRule } from "../../../core/slots/ColorRule.js";
import { DEFAULT_FORMATTER } from "../../shared/constants.js";
import {
  defaultToolbarInternalConfig
} from "../../shared/types/meter-bar-toolbar.js";
import { defaultConfig } from "../types/default.js";
const getTooltipConfig = (slots, consumerConfig) => {
  const consumerTooltip = consumerConfig.tooltip !== void 0 ? { hidden: !consumerConfig.tooltip } : void 0;
  return { ...defaultConfig.tooltip, ...consumerTooltip, ...slots.tooltip };
};
const getToolbarConfig = (toolbarSlotsConfig, toolbarConsumerConfig) => {
  const configuredToolbar = toolbarSlotsConfig ?? toolbarConsumerConfig;
  return { ...defaultToolbarInternalConfig, ...configuredToolbar };
};
const buildMeterBarChartConfig = (props, consumerConfig, slots) => {
  const thresholds = combineThresholds(
    consumerConfig.thresholdsIndicators,
    slots.thresholds
  );
  const size = props.size ?? consumerConfig.size ?? defaultConfig.size;
  const min = props.min ?? consumerConfig.min ?? defaultConfig.min;
  const max = props.max ?? consumerConfig.max ?? defaultConfig.max;
  const tooltip = getTooltipConfig(slots, consumerConfig);
  const formatter = props.formatter ?? consumerConfig.formatter ?? DEFAULT_FORMATTER;
  const unit = props.unit ?? consumerConfig.unit;
  const toolbar = getToolbarConfig(slots.toolbar, consumerConfig.toolbar);
  const colorRules = (consumerConfig.colorRules ?? []).filter(isValidColorRule).concat(slots.colorRules ?? []);
  const loading = props.loading ?? consumerConfig.loading ?? defaultConfig.loading;
  return {
    thresholds,
    thresholdsIndicators: thresholds,
    size,
    loading,
    min,
    max,
    tooltip,
    labelSlot: slots.label,
    valueSlot: slots.value,
    minSlot: slots.min,
    maxSlot: slots.max,
    formatter,
    unit,
    colorRules,
    toolbar
  };
};
const combineThresholds = (thresholdConfig, thresholdSlots) => {
  const slotsData = thresholdSlots?.data ? thresholdSlots.data : [];
  const consumerData = thresholdConfig?.data?.map(applyThresholdDefaults) ?? [];
  return {
    data: slotsData.concat(consumerData),
    legend: thresholdSlots?.legend ?? thresholdConfig?.legend ?? defaultConfig.thresholdIndicators.legend
  };
};
export {
  buildMeterBarChartConfig
};
