// @dynatrace/strato-components/esm/charts/core/components/legend/categorical/ChartLegend.js
import { jsx } from "react/jsx-runtime";
import { useRef } from "react";
import { legendContainer } from "./ChartLegend.sty.js";
import { ChartLegendGrid } from "./components/ChartLegendGrid.js";
import { ChartVirtualizedLegendGrid } from "./components/ChartVirtualizedLegendGrid.js";
import { LegendGridSettingsContext } from "./context/LegendGridSettingsContext.js";
import { useLegendInteractions } from "./hooks/useLegendInteractions.js";
import { buildGridSettings } from "./utils/legend-grid.utils.js";
import { DEFAULT_TRUNCATION_MODE, ORIENTATION } from "../../../constants.js";
import { LegendContext } from "../shared/context/Legend.context.js";
import { useLegendSize } from "../shared/context/LegendSizeContext.js";
const ChartLegend = (props) => {
  const {
    items,
    orientation = ORIENTATION.Horizontal,
    onHover,
    onClick,
    onSeriesAction,
    interactionsDisabled = false,
    truncationMode = DEFAULT_TRUNCATION_MODE,
    style
  } = props;
  const ref = useRef(null);
  const { width, height } = useLegendSize();
  const { onItemHover, onLegendIconClick, onLegendNameClick, hiddenItems } = useLegendInteractions(items, { onClick, onHover });
  const gridSettings = buildGridSettings(items, width, height);
  return /* @__PURE__ */ jsx(
    "div",
    {
      ref,
      "data-testid": "chart-legend",
      className: legendContainer,
      style: {
        height,
        width,
        flexDirection: orientation === ORIENTATION.Vertical ? "row" : "column",
        ...style
      },
      children: /* @__PURE__ */ jsx(
        LegendContext.Provider,
        {
          value: {
            hiddenItems,
            truncationMode,
            onSeriesAction,
            onLegendIconClick,
            onItemHover,
            onLegendNameClick,
            interactionsDisabled
          },
          children: /* @__PURE__ */ jsx(LegendGridSettingsContext.Provider, { value: gridSettings, children: gridSettings.hasVirtual ? /* @__PURE__ */ jsx(
            ChartVirtualizedLegendGrid,
            {
              items,
              orientation
            }
          ) : /* @__PURE__ */ jsx(ChartLegendGrid, { items, orientation }) })
        }
      )
    }
  );
};
ChartLegend["displayName"] = "ChartLegend";
export {
  ChartLegend
};
