// @dynatrace/strato-components/charts/pie/hooks/useSliceHandlers.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useSliceHandlers_exports = {};
__export(useSliceHandlers_exports, {
  useSliceHandlers: () => useSliceHandlers
});
module.exports = __toCommonJS(useSliceHandlers_exports);
var import_TooltipContext = require("../../core/components/overlay-tooltip/contexts/TooltipContext.js");
var import_useTooltipActions = require("../../core/components/overlay-tooltip/useTooltipActions.js");
var import_useEventErrorHandler = require("../../core/hooks/useEventErrorHandler.js");
var import_get_tooltip_content = require("../utils/get-tooltip-content.js");
const buildTooltipData = (sliceId, data, handlerCatch, formatter) => {
  const tooltipContent = (0, import_get_tooltip_content.getTooltipContent)(
    sliceId,
    data,
    handlerCatch,
    formatter
  );
  return {
    metadata: {
      sliceId: data.id
    },
    geometry: "pie",
    series: [
      {
        name: tooltipContent.name,
        value: tooltipContent.absoluteValue,
        color: tooltipContent.color,
        relativeValue: tooltipContent.relativeValue,
        relativeValueUnit: "%"
      }
    ]
  };
};
function useSliceHandlers(sliceId, data, formatter) {
  const handlerCatch = (0, import_useEventErrorHandler.useEventErrorHandler)();
  const tooltipState = (0, import_TooltipContext.useTooltip)();
  const { hover, pin, unpin, reset } = (0, import_useTooltipActions.useTooltipActions)();
  const onPointerMoveHandler = (event) => {
    if (tooltipState.getState().pinned || sliceId === "slice-thinner-slices") {
      return;
    }
    const tooltipData = buildTooltipData(
      sliceId,
      data,
      handlerCatch,
      formatter
    );
    hover(event, tooltipData);
  };
  const onMouseClickHandler = (event) => {
    event.stopPropagation();
    if (sliceId === "slice-thinner-slices") {
      reset();
    } else {
      const tooltipData = buildTooltipData(
        sliceId,
        data,
        handlerCatch,
        formatter
      );
      if (tooltipState.getState().pinned) {
        unpin();
        hover(event, tooltipData);
      } else {
        pin(event, tooltipData);
      }
    }
  };
  return {
    onPointerMoveHandler,
    onMouseClickHandler
  };
}
