// @dynatrace/strato-components/esm/editors/base-code-editor/highlight/highlight.extension.js
import { Prec, RangeSetBuilder } from "@codemirror/state";
import {
  Decoration,
  ViewPlugin
} from "@codemirror/view";
const highlight = Decoration.mark({
  attributes: {
    style: "animation: none"
  },
  class: "cm-highlight"
});
function highlightUpdatedLines(view, previousValue) {
  const builder = new RangeSetBuilder();
  const previousSplittedValue = previousValue.split("\n");
  if (previousValue) {
    for (const { from, to } of view.visibleRanges) {
      for (let pos = from; pos <= to; ) {
        const line = view.state.doc.lineAt(pos);
        if (previousSplittedValue[line.number - 1] !== line.text) {
          builder.add(line.from, line.to, highlight);
        }
        pos = line.to + 1;
      }
    }
  }
  return builder.finish();
}
const highlightExtension = (previousValue) => {
  const highlightLines = ViewPlugin.fromClass(
    class {
      decorations;
      constructor(view) {
        this.decorations = highlightUpdatedLines(view, previousValue);
        requestAnimationFrame(() => {
          view?.dom.querySelectorAll(".cm-highlight").forEach((n) => {
            n.style.animation = "";
          });
        });
      }
      /** Run update if user changed input/focus */
      update(update) {
        if (update.docChanged || update.focusChanged) {
          this.decorations = new RangeSetBuilder().finish();
        }
      }
    },
    {
      decorations: (v) => v.decorations
    }
  );
  return Prec.default(highlightLines);
};
export {
  highlightExtension
};
