// @dynatrace/strato-components/esm/forms/select/Select.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import {
  forwardRef,
  useCallback,
  useEffect,
  useImperativeHandle,
  useMemo,
  useReducer,
  useRef,
  useState
} from "react";
import ReactSelect, { createFilter } from "react-select";
import { components } from "./components/components.js";
import { SelectMenuContext } from "./contexts/SelectMenuContext.js";
import { useSelectOptions, _isOption } from "./hooks/use-select-options.js";
import { useSelectSlots } from "./hooks/use-select-slots.js";
import { useSelectValue } from "./hooks/use-select-value.js";
import { SelectContent } from "./SelectContent.js";
import { SelectCustomTrigger } from "./SelectCustomTrigger.js";
import { SelectDisplayValue } from "./SelectDisplayValue.js";
import {
  SelectEmptyState
} from "./SelectEmptyState.js";
import { SelectFilter } from "./SelectFilter.js";
import { SelectGroup, SelectGroupLabel } from "./SelectGroup.js";
import { SelectOption } from "./SelectOption.js";
import { SelectPrefix } from "./SelectPrefix.js";
import { SelectSuffix } from "./SelectSuffix.js";
import { SelectTrigger } from "./SelectTrigger.js";
import {
  reducer,
  SelectActions
} from "./state/select-menu-state-reducer.js";
import * as styles from "./styles/Select.sty.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useControllableState } from "../../core/hooks/useControllableState.js";
import { useFormControlState } from "../../core/hooks/useFormControlState.js";
import { createStratoLogger } from "../../core/sdk/sdk-logger.js";
import { mergeProps } from "../../core/utils/merge-props.js";
import { withFormFieldMessageContextProviderWithGenerics } from "../forms-core/components/form-field/FormFieldMessageContextProvider.js";
import { useFormFieldMessagesContext } from "../forms-core/hooks/useFormFieldMessagesContext.js";
import { useFormFieldValidation } from "../forms-core/hooks/useFormFieldValidation.js";
import { useSyncedFormControlProps } from "../forms-core/hooks/useSyncedFormControlProps.js";
import { FormFieldMessagesAction } from "../forms-core/state/formfield-messages-state-reducer.js";
const COMPONENT_NAME = "Select";
function getNewFocusedOption(options, focusedOptionValue) {
  for (const optionOrGroup of options) {
    if (_isOption(optionOrGroup)) {
      if (optionOrGroup.value === focusedOptionValue) {
        return optionOrGroup;
      }
    } else {
      for (const option of optionOrGroup.options) {
        if (_isOption(option) && option.value === focusedOptionValue) {
          return option;
        }
      }
    }
  }
  return null;
}
function SelectFwd(props, forwardedRef) {
  const {
    multiple = false,
    disabled: propsDisabled,
    controlState,
    clearable = multiple,
    name,
    id: propId,
    readOnly: propReadOnly,
    children,
    value: propsValue,
    defaultValue: propsDefaultValue,
    onChange: propsOnChange,
    onFocus,
    onBlur,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "aria-label": ariaLabel,
    "aria-labelledby": ariaLabelledBy,
    "aria-describedby": ariaDescribedBy,
    "aria-details": ariaDetails,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "aria-disabled": ariaDisabledProp,
    onOpenChange,
    required: propsRequired,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
  const slots = useSelectSlots(children);
  const stratoLogger = useMemo(
    () => createStratoLogger.bind({
      name,
      id: propId,
      ariaLabel,
      dataTestId
    })(COMPONENT_NAME),
    [name, propId, ariaLabel, dataTestId]
  );
  const {
    disableFiltering,
    value: filterValueProp,
    defaultValue: filterDefaultValueProp,
    onChange: filterOnChangeProp,
    ...remainingFilterProps
  } = slots.Filter?.props ?? {};
  const {
    width: triggerWidthProp,
    placeholder: triggerPlaceholderProp,
    ...remainingTriggerProps
  } = slots.Trigger?.props ?? {};
  const {
    width: contentWidthProp,
    loading: contentLoadingProp,
    showSelectedOptionsFirst: contentShowSelectedOptionsFirstProp,
    ...remainingContentProps
  } = slots.Content?.props ?? {};
  const {
    optionsOrGroups: options,
    optionsHavePrefix,
    allOptions
  } = useSelectOptions(slots.Content, stratoLogger);
  const { value, defaultValue } = useSelectValue({
    value: propsValue,
    defaultValue: propsDefaultValue,
    options,
    multiple: multiple === true
  });
  const hiddenSelectRef = useRef(null);
  const [triggerRef, setTriggerRef] = useState(null);
  const { disabled, required, id, ariaDisabled, readOnly } = useSyncedFormControlProps({
    propDisabled: propsDisabled,
    propRequired: propsRequired,
    propId,
    propAriaDisabled: ariaDisabledProp,
    propReadOnly
  });
  const { onBlur: validationOnBlur } = useFormFieldValidation(hiddenSelectRef, {
    required
  });
  const menuOpenReducer = useCallback(
    (state, action) => {
      if (readOnly) {
        return state;
      }
      const newState = reducer(state, action);
      if (state !== newState) {
        queueMicrotask(() => {
          newState === false && !controlState && validationOnBlur();
          onOpenChange?.(newState);
        });
      }
      return newState;
    },
    [readOnly, controlState, validationOnBlur, onOpenChange]
  );
  const [menuOpen, dispatch] = useReducer(menuOpenReducer, false);
  const [isFocusWithin, setIsFocusWithin] = useState(false);
  const [filterValue, setFilterValue] = useControllableState({
    value: filterValueProp,
    defaultValue: filterDefaultValueProp,
    onChange: filterOnChangeProp
  });
  const { state: messagesContextState, dispatch: messagesContextDispatch } = useFormFieldMessagesContext();
  const selectRef = useRef(null);
  const hiddenTrigger = useMemo(
    () => triggerRef?.querySelector("input"),
    [triggerRef]
  );
  const [focusOnOption, setFocusOnOption] = useState(true);
  useEffect(() => {
    if (disableFiltering === true && (filterValue !== void 0 || filterValue !== "")) {
      setFilterValue("");
    }
  }, [disableFiltering]);
  const filterConfig = {
    // Filter by both visible label and option value/textValue.
    // This allows searching by value (e.g. "green") when the rendered label differs.
    stringify: (option) => {
      const searchValue = option.data.textValue ?? option.value;
      if (searchValue && searchValue !== option.label) {
        return `${option.label} ${searchValue}`;
      }
      return `${option.label}`;
    }
  };
  const handleChange = (option) => {
    requestAnimationFrame(() => {
      const nativeEvent = new Event("change", {
        bubbles: true,
        cancelable: true
      });
      hiddenSelectRef?.current?.dispatchEvent(nativeEvent);
    });
    if (multiple === false) {
      setFilterValue("");
    }
    if (!propsOnChange) {
      return;
    }
    if (multiple === false) {
      if (option === value) {
        return;
      }
      if (option === null) {
        propsOnChange?.(null);
      } else {
        propsOnChange?.(
          option?.value
        );
      }
    }
    if (multiple === true) {
      const selectedOptionValues = option.map((individualOption) => individualOption.value);
      propsOnChange?.(
        selectedOptionValues
      );
    }
  };
  const filterRef = useRef(null);
  const clearSelectionRef = useRef(null);
  const containerRef = useRef(null);
  const contentRef = useRef(null);
  const menuListFunctionRef = useRef(null);
  const handleArrowKeyOptionFocus = (key) => {
    if (key !== "ArrowUp" && key !== "ArrowDown") {
      return;
    }
    requestAnimationFrame(() => {
      if (!contentRef.current) {
        return;
      }
      const container = contentRef.current;
      const focusedElement = container.querySelector(
        ".strato-select-option--focus"
      );
      if (focusedElement) {
        focusedElement.scrollIntoView({
          block: "nearest"
        });
      } else if (container) {
        if (key === "ArrowDown") {
          container.scrollTop = 0;
        } else if (menuListFunctionRef.current) {
          menuListFunctionRef.current.scrollToBottom();
        }
      }
    });
  };
  useImperativeHandle(
    forwardedRef,
    () => ({
      element: containerRef.current,
      triggerRef,
      inputRef: hiddenSelectRef.current,
      focus: () => {
        hiddenTrigger?.focus();
      },
      open: () => {
        hiddenTrigger?.focus();
        dispatch({
          type: SelectActions.OPEN
        });
      },
      close: () => dispatch({
        type: SelectActions.CLOSE
      }),
      validate: () => {
        queueMicrotask(() => hiddenSelectRef.current?.reportValidity());
        messagesContextDispatch?.({
          type: FormFieldMessagesAction.SETISTOUCHED,
          payload: { isTouched: true }
        });
      }
    }),
    [hiddenTrigger, messagesContextDispatch, triggerRef]
  );
  if (multiple && propsOnChange) {
    const select = selectRef.current;
    if (select?.state.focusedOption) {
      const focusedOptionValue = select.state.focusedOption.value;
      const newFocusedOption = getNewFocusedOption(options, focusedOptionValue);
      select.state.focusedOption = newFocusedOption;
    }
  }
  const [hasFormControlError] = useFormControlState(controlState);
  const showError = controlState ? hasFormControlError : messagesContextState.showError;
  const contextMemo = useMemo(
    () => ({ state: menuOpen, dispatch }),
    [menuOpen, dispatch]
  );
  const onBlurHandler = (event) => {
    if (
      // The relatedTarget is null when the focus leaves the browser and we want to still trigger the onBlur in this case.
      !event.relatedTarget || event.relatedTarget !== hiddenTrigger && event.relatedTarget !== clearSelectionRef.current && event.relatedTarget !== filterRef.current?.inputRef
    ) {
      if (!menuOpen && isFocusWithin) {
        setIsFocusWithin(false);
        onBlur?.(event);
      }
    }
  };
  return /* @__PURE__ */ jsx(SelectMenuContext.Provider, { value: contextMemo, children: /* @__PURE__ */ jsx(
    "div",
    {
      ref: containerRef,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      className: clsx(
        styles.selectWrapper({
          fullWidth: triggerWidthProp === "full" || triggerWidthProp === "100%"
        }),
        consumerClassName
      ),
      style: consumerStyle,
      ...mergeProps(restProps, behavioralTrackingProps),
      "data-dt-component": COMPONENT_NAME,
      children: /* @__PURE__ */ jsx(
        ReactSelect,
        {
          instanceId: id,
          selectInputId: id,
          className: clsx(styles.select, styles.clearReactSelectZIndex, {
            [styles.validAndEnabled]: !disabled && !ariaDisabled && !showError
          }),
          unstyled: true,
          options,
          components,
          value,
          defaultValue,
          onChange: (value2) => (
            // Need to cast this here as React.Select handles the value still as unknown.
            // TODO: Maybe we can go to the JSX generic?
            handleChange(
              value2
            )
          ),
          onBlur: onBlurHandler,
          hiddenSelectName: name,
          ref: selectRef,
          pageSize: 5,
          "aria-label": ariaLabel,
          "aria-labelledby": ariaLabelledBy,
          "aria-describedby": ariaDescribedBy,
          "aria-details": ariaDetails,
          "aria-disabled": ariaDisabled,
          hiddenSelectRef,
          isSearchable: false,
          isMulti: multiple ?? false,
          hideSelectedOptions: false,
          closeMenuOnSelect: multiple === false,
          isClearable: false,
          isDisabled: disabled,
          isRequired: required,
          triggerWidth: triggerWidthProp,
          contentWidth: contentWidthProp ?? "trigger",
          placeholderText: triggerPlaceholderProp,
          displayValue: slots.DisplayValue ?? void 0,
          prefix: slots.Prefix ?? void 0,
          customTrigger: slots.CustomTrigger ?? void 0,
          triggerRef,
          setTriggerRef,
          optionsHavePrefix,
          menuIsOpen: menuOpen,
          filterOption: (option, inputValue) => {
            if (disableFiltering === true) {
              return true;
            }
            const filterFunction = createFilter(filterConfig);
            return filterFunction(
              option,
              inputValue
            );
          },
          customFilter: slots.Filter ?? void 0,
          inputValue: filterValue,
          onInputChange: (newValue, actionmeta) => {
            if (actionmeta.action === "input-change") {
              setFilterValue(newValue);
            }
          },
          emptyState: slots.EmptyState ?? void 0,
          controlState,
          loading: contentLoadingProp,
          clearable,
          allOptions,
          tabSelectsValue: false,
          focusOnOption,
          contentRef,
          menuListFunctionRef,
          filterRef,
          clearSelectionRef,
          showSelectedOptionsFirst: contentShowSelectedOptionsFirstProp,
          componentProps: {
            trigger: {
              className: slots.Trigger?.props.className,
              style: { ...slots.Trigger?.props.style },
              readOnly: !!readOnly,
              ariaDisabled,
              ...remainingTriggerProps
            },
            filter: {
              className: slots.Filter?.props.className,
              style: slots.Filter?.props.style,
              ...remainingFilterProps
            },
            content: {
              className: slots.Content?.props.className,
              style: slots.Content?.props.style,
              ...remainingContentProps
            },
            emptyState: {
              ...slots.EmptyState?.props
            }
          },
          onFocus: (event) => {
            if (!isFocusWithin) {
              onFocus?.(event);
              setIsFocusWithin(true);
            }
          },
          onKeyDown: (event) => {
            if (!menuOpen) {
              if (event.key === "Enter" || event.key === " ") {
                dispatch({
                  type: SelectActions.SELECTFOCUS
                });
              }
              return true;
            }
            if (event.key === "Escape") {
              dispatch({
                type: SelectActions.SELECTFOCUS
              });
              hiddenTrigger?.focus();
            }
            if (filterRef.current || clearSelectionRef.current) {
              if (document.activeElement === filterRef.current?.inputRef) {
                if (event.key === "Backspace" && filterRef.current.inputRef?.value === "") {
                  event.preventDefault();
                  event.stopPropagation();
                }
                if (event.key === "Tab") {
                  if (!clearSelectionRef.current || event.shiftKey) {
                    event.preventDefault();
                    event.stopPropagation();
                    hiddenTrigger?.focus();
                  } else {
                    clearSelectionRef.current && setFocusOnOption(false);
                  }
                }
                if (event.key === "Enter" && !props.multiple) {
                  dispatch({
                    type: SelectActions.SELECTFOCUS
                  });
                  hiddenTrigger?.focus();
                }
                if (event.key === "Enter" && contentLoadingProp) {
                  event.preventDefault();
                  event.stopPropagation();
                }
                if (event.key === " " && !filterValue) {
                  setFilterValue(" ");
                  event.preventDefault();
                  event.stopPropagation();
                }
                handleArrowKeyOptionFocus(event.key);
                return true;
              }
              if (document.activeElement === clearSelectionRef.current) {
                setFocusOnOption(false);
                if (event.key === "ArrowDown" || event.key === "ArrowUp") {
                  event.preventDefault();
                }
                if (event.key === "Tab" && !(event.shiftKey && filterRef.current)) {
                  event.preventDefault();
                  event.stopPropagation();
                  hiddenTrigger?.focus();
                  setFocusOnOption(true);
                }
                if (event.key === "Tab" && event.shiftKey && filterRef.current) {
                  setFocusOnOption(true);
                }
                return false;
              }
              try {
                if (document.activeElement && document.getElementById(document.activeElement.id)) {
                  if (event.key === "Tab") {
                    event.preventDefault();
                    event.stopPropagation();
                    if (!event.shiftKey) {
                      if (filterRef.current) {
                        filterRef.current.inputRef?.focus();
                        return false;
                      }
                      if (clearSelectionRef.current) {
                        clearSelectionRef.current.focus();
                        setFocusOnOption(false);
                        return false;
                      }
                    } else {
                      if (clearSelectionRef.current) {
                        clearSelectionRef.current.focus();
                        setFocusOnOption(false);
                        return false;
                      }
                      if (filterRef.current) {
                        filterRef.current.inputRef?.focus();
                        return false;
                      }
                    }
                  }
                  handleArrowKeyOptionFocus(event.key);
                  if (contentLoadingProp) {
                    event.preventDefault();
                    event.stopPropagation();
                  }
                }
              } catch (error) {
                console.log(
                  "Invalid query selector on the Select trigger",
                  error
                );
              }
              return false;
            }
            try {
              if (document.activeElement && document.getElementById(document.activeElement.id)) {
                if (event.key === "Tab") {
                  dispatch({
                    type: SelectActions.SELECTFOCUS
                  });
                } else {
                  handleArrowKeyOptionFocus(event.key);
                }
              }
            } catch (error) {
              console.log(
                "Invalid query selector on the Select trigger",
                error
              );
            }
            setFocusOnOption(true);
          }
        }
      )
    }
  ) });
}
const SelectForwardRef = forwardRef(SelectFwd);
const SelectRoot = withFormFieldMessageContextProviderWithGenerics(SelectForwardRef);
const Select = Object.assign(SelectRoot, {
  Content: SelectContent,
  Option: SelectOption,
  Group: SelectGroup,
  GroupLabel: SelectGroupLabel,
  Trigger: SelectTrigger,
  DisplayValue: SelectDisplayValue,
  Prefix: SelectPrefix,
  Suffix: SelectSuffix,
  CustomTrigger: SelectCustomTrigger,
  Filter: SelectFilter,
  EmptyState: SelectEmptyState
});
Select.displayName = COMPONENT_NAME;
export {
  Select
};
