// @dynatrace/strato-components/esm/content/avatar-group/AvatarGroup.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import {
  forwardRef,
  Children,
  useMemo,
  cloneElement
} from "react";
import { useIntl } from "react-intl";
import { additionalAvatarWrapperCSS } from "./AvatarGroup.sty.js";
import { Item } from "./Item.js";
import { Button } from "../../buttons/button/Button.js";
import { useAriaLabelingProps } from "../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useMergeRefs } from "../../core/hooks/useMergeRefs.js";
import { useRovingFocus } from "../../core/hooks/useRovingFocus.js";
import { useSafeForwardProps } from "../../core/hooks/useSafeForwardProps.js";
import { focusRingStatic } from "../../core/styles/focusRingStatic.sty.js";
import { isSlot } from "../../core/utils/is-slot.js";
import { Flex } from "../../layouts/flex/Flex.js";
import { Menu } from "../../navigation/menu/Menu.js";
const COMPONENT_NAME = "AvatarGroup";
const MAX_AVATARGROUP_ITEMS = 5;
const RootAvatarGroup = forwardRef(
  (props, forwardedRef) => {
    const {
      children,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
    const intl = useIntl();
    const ariaLabelingProps = useAriaLabelingProps(restProps);
    const forwardProps = useSafeForwardProps(restProps, ariaLabelingProps);
    const { "aria-label": ariaLabel, ...remainingAriaLabelingProps } = ariaLabelingProps;
    const { containerRef: rovingContainerRef } = useRovingFocus(
      {
        initialIndex: void 0,
        direction: "horizontal",
        focusableChildrenSelector: ".rovingFocusElement"
      }
    );
    const [displayValues, additionalValues] = useMemo(() => {
      const childrenArray = Children.toArray(children);
      let displayValues2;
      const additionalValues2 = [];
      if (childrenArray.length > MAX_AVATARGROUP_ITEMS) {
        displayValues2 = childrenArray.splice(0, MAX_AVATARGROUP_ITEMS);
        for (let i = 0; i < childrenArray.length; i++) {
          const child = childrenArray[i];
          if (isSlot(child, Item)) {
            additionalValues2.push(
              cloneElement(
                /* @__PURE__ */ jsx(
                  Item,
                  {
                    abbreviation: child.props.abbreviation,
                    tooltipText: child.props.tooltipText
                  }
                ),
                {
                  ...child.props,
                  key: i,
                  isCollapsed: true
                }
              )
            );
          }
        }
      } else {
        displayValues2 = childrenArray;
      }
      return [displayValues2, additionalValues2];
    }, [children]);
    const rootRef = useMergeRefs([forwardedRef, rovingContainerRef]);
    const formatMessageAriaLabel = intl.formatMessage(
      {
        id: "4ER0rfN8lI3HtaFd",
        defaultMessage: "Group of Avatars. {hiddenAvatars} hidden, {totalAvatars} total users",
        description: "Label for the collapsed Avatars."
      },
      {
        hiddenAvatars: additionalValues.length,
        totalAvatars: additionalValues.length + displayValues.length
      }
    );
    return displayValues.length ? /* @__PURE__ */ jsxs(
      Flex,
      {
        gap: 4,
        alignItems: "center",
        ref: rootRef,
        className: clsx(focusRingStatic(), consumerClassName),
        style: consumerStyle,
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        "aria-label": ariaLabel ?? formatMessageAriaLabel,
        ...remainingAriaLabelingProps,
        ...behavioralTrackingProps,
        ...forwardProps,
        "data-dt-component": COMPONENT_NAME,
        children: [
          /* @__PURE__ */ jsx(Flex, { alignItems: "center", gap: 0, children: displayValues }),
          additionalValues.length > 0 && /* @__PURE__ */ jsxs(Menu, { children: [
            /* @__PURE__ */ jsx(Menu.Trigger, { children: /* @__PURE__ */ jsxs(
              Button,
              {
                size: "condensed",
                variant: "emphasized",
                className: clsx(
                  "rovingFocusElement",
                  consumerClassName,
                  additionalAvatarWrapperCSS
                ),
                children: [
                  "+",
                  additionalValues.length
                ]
              }
            ) }),
            /* @__PURE__ */ jsx(
              Menu.Content,
              {
                "data-dtrum-mask": dataDtrumMask,
                "data-dtrum-allow": dataDtrumAllow,
                children: additionalValues
              }
            )
          ] })
        ]
      }
    ) : null;
  }
);
RootAvatarGroup.displayName = COMPONENT_NAME;
const AvatarGroup = Object.assign(RootAvatarGroup, {
  Item
});
export {
  AvatarGroup
};
