// @dynatrace/strato-components/esm/content/avatar-group/Item.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import {
  forwardRef
} from "react";
import { avatarGroupItemWrapperCSS } from "./AvatarGroup.sty.js";
import { Menu } from "../../navigation/menu/Menu.js";
import { Tooltip } from "../../overlays/tooltip/Tooltip.js";
import { Avatar } from "../avatar/Avatar.js";
const Item = /* @__PURE__ */ forwardRef(
  (props, ref) => {
    const {
      size = "default",
      abbreviation,
      disabled = false,
      tooltipText,
      as,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...remainingProps
    } = props;
    const isCollapsed = remainingProps.isCollapsed;
    return !isCollapsed ? /* @__PURE__ */ jsx(
      "div",
      {
        className: clsx(consumerClassName, avatarGroupItemWrapperCSS),
        style: consumerStyle,
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        children: /* @__PURE__ */ jsx(
          Tooltip,
          {
            text: tooltipText,
            fallbackPlacements: [
              "top",
              "bottom",
              "top-start",
              "top-end",
              "bottom-start",
              "bottom-end",
              "right",
              "left",
              "right-start",
              "right-end",
              "left-start",
              "left-end"
            ],
            "data-dtrum-mask": dataDtrumMask,
            "data-dtrum-allow": dataDtrumAllow,
            children: /* @__PURE__ */ jsx(
              Avatar,
              {
                abbreviation,
                ref,
                size,
                as,
                className: "rovingFocusElement",
                disabled,
                ...remainingProps
              }
            )
          }
        )
      }
    ) : /* @__PURE__ */ jsx(
      Menu.Item,
      {
        "data-testid": dataTestId,
        onSelect: (() => null),
        className: clsx("rovingFocusElement", consumerClassName),
        as,
        disabled,
        style: consumerStyle,
        ...remainingProps,
        children: /* @__PURE__ */ jsx(Avatar, { abbreviation, size: "small", disabled, children: /* @__PURE__ */ jsx(Avatar.Label, { children: tooltipText }) })
      }
    );
  }
);
Item.displayName = "AvatarGroup.Item";
export {
  Item
};
