// @dynatrace/strato-components/esm/tables/utils/log-content-highlighting.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import { useMemo, Fragment as Fragment2 } from "react";
import { ErrorBoundary } from "react-error-boundary";
import { HighlightExtended } from "../../content/code-snippet/HighlightExtended.js";
import { getStratoLogger } from "../../core/sdk/sdk-logger.js";
import { textStyle as textStyleCSS } from "../../styles/textStyle.sty.js";
const ELLIPSIS_CHARACTER = "\u2026";
const MAX_TOKENIZATION_LENGTH = 1e3;
const DEFAULT_TRUNCATION_LENGTH = MAX_TOKENIZATION_LENGTH;
function useFormatLogContent(logContent, options) {
  const stratoLogger = getStratoLogger("DataTable");
  return useMemo(() => {
    if (logContent !== null && logContent !== void 0) {
      const lineWrap = options?.lineWrap ?? true;
      let logContentValue = logContent;
      if (!lineWrap) {
        logContentValue = logContent.replace(/[\n\r]/g, " ");
      }
      const truncationLimit = getTruncationLimit(options?.truncationLimit);
      const highlightIndex = Math.min(
        MAX_TOKENIZATION_LENGTH,
        truncationLimit,
        // in case truncation limit is lower than 1000
        logContentValue.length
        // in case log line is shorter than 1000 chars
      );
      let logContentSubstrings = {
        // max. first 1000 characters are tokenized and highlighted
        highlight: logContentValue.substring(0, highlightIndex),
        // remaining characters that are neither tokenized nor highlighted to enhance performance
        noHighlight: logContentValue.substring(
          highlightIndex,
          Math.min(truncationLimit, logContentValue.length)
        )
      };
      if (truncationLimit < logContentValue.length) {
        logContentSubstrings = ellipsize(logContentSubstrings);
      }
      const noHighlightLines = logContentSubstrings.noHighlight.length > 0 ? logContentSubstrings.noHighlight.split(/\r?\n/) : [];
      return /* @__PURE__ */ jsx(
        ErrorBoundary,
        {
          fallback: /* @__PURE__ */ jsx("span", { children: logContentValue.substring(0, truncationLimit) }),
          onError: (error, info) => {
            stratoLogger.error(error.message, error, {
              logContentValue,
              info
            });
          },
          children: /* @__PURE__ */ jsx(
            HighlightExtended,
            {
              language: "log",
              code: logContentSubstrings.highlight,
              children: ({ tokens, getLineProps, getTokenProps, className, style }) => /* @__PURE__ */ jsx(Fragment, { children: tokens.map((line, i) => {
                const { className: lineClassName, ...remainingLineProps } = getLineProps({ line });
                return /* @__PURE__ */ jsxs(
                  "div",
                  {
                    ...remainingLineProps,
                    className: clsx(
                      className,
                      textStyleCSS({ fontStyle: "code" })
                    ),
                    style: { ...style, width: "100%" },
                    children: [
                      line.map((token, j) => /* @__PURE__ */ jsx("span", { ...getTokenProps({ token }) }, j)),
                      /* render the remaining content that is not highlighted
                         in the same `token-line` after the last token  */
                      i === tokens.length - 1 ? noHighlightLines.map((line2, j) => /* @__PURE__ */ jsxs(Fragment2, { children: [
                        /* @__PURE__ */ jsx("span", { children: line2 }),
                        j < noHighlightLines.length - 1 ? /* @__PURE__ */ jsx("br", {}) : null
                      ] }, `${i}-${j}`)) : null
                    ]
                  },
                  i
                );
              }) })
            },
            logContentSubstrings.highlight
          )
        }
      );
    }
    return null;
  }, [logContent, options?.lineWrap, options?.truncationLimit, stratoLogger]);
}
function getTruncationLimit(truncationLimit) {
  if (truncationLimit === void 0) {
    return DEFAULT_TRUNCATION_LENGTH;
  }
  if (truncationLimit === false) {
    return Infinity;
  }
  return truncationLimit;
}
function ellipsize(content) {
  let { highlight, noHighlight } = content;
  const trimmedNoHighlight = noHighlight.trimEnd();
  if (trimmedNoHighlight.length > 0) {
    noHighlight = trimmedNoHighlight + ELLIPSIS_CHARACTER;
  } else {
    const trimmedHighlight = content.highlight.trimEnd();
    if (trimmedHighlight.length > 0) {
      highlight = trimmedHighlight + ELLIPSIS_CHARACTER;
    }
  }
  return { highlight, noHighlight };
}
function LogContentHighlight(props) {
  const { logContent, lineWrap, truncationLimit } = props;
  return useFormatLogContent(logContent, { lineWrap, truncationLimit });
}
export {
  LogContentHighlight
};
