// @dynatrace/strato-components/filters/filter-field/hooks/recent-and-pinned/usePersistedFilter.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var usePersistedFilter_exports = {};
__export(usePersistedFilter_exports, {
  usePersistedFilter: () => usePersistedFilter,
  usePersistedFilterMutation: () => usePersistedFilterMutation
});
module.exports = __toCommonJS(usePersistedFilter_exports);
var import_react_query = require("@tanstack/react-query");
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_app_environment = require("@dynatrace-sdk/app-environment");
var import_client_state = require("@dynatrace-sdk/client-state");
var import_useRecentFiltersStore = require("./useRecentFiltersStore.js");
var import_SavedAndRecentQueryClientContext = require("../../contexts/SavedAndRecentQueryClientContext.js");
var import_useFilterFieldLoggerContext = require("../useFilterFieldLoggerContext.js");
const currentUser = (0, import_app_environment.getCurrentUserDetails)();
function usePersistedFilter(name, key, saveFn) {
  const queryClient = (0, import_react.useContext)(import_SavedAndRecentQueryClientContext.RecentAndSavedFiltersQueryClientContext);
  const setCanPin = (0, import_useRecentFiltersStore.useRecentAndPinnedFiltersStore)(
    (state) => state.setCanPin,
    name
  );
  const { logger } = (0, import_useFilterFieldLoggerContext.useFilterFieldLoggerContext)();
  return (0, import_react_query.useQuery)(
    {
      enabled: name !== null,
      queryKey: [key],
      throwOnError: false,
      queryFn: async ({ signal }) => {
        let savedFilters;
        try {
          savedFilters = await import_client_state.stateClient.getUserAppState({
            key,
            abortSignal: signal
          });
        } catch (error) {
          if ((0, import_client_state.isUnauthorized)(error) || (0, import_client_state.isForbidden)(error)) {
            logger?.warn(
              "FilterField pinned and recent error - permissions not met",
              error
            );
            setCanPin(false);
          } else if (signal.aborted) {
          } else if (!(0, import_client_state.isNotFound)(error)) {
            logger?.warn("FilterField pinned and recent error", error);
          }
          return [];
        }
        const parsedValue = JSON.parse(savedFilters.value, (key2, value) => {
          if (key2 === "timeUsed") {
            return new Date(value);
          }
          return value;
        });
        for (const parsedFilter of parsedValue) {
          if (!parsedFilter.timeUsed) {
            parsedFilter.timeUsed = /* @__PURE__ */ new Date();
          }
        }
        saveFn(parsedValue);
        return parsedValue;
      }
    },
    queryClient
  );
}
function usePersistedFilterMutation(key, name, actionType) {
  const queryClient = (0, import_react.useContext)(import_SavedAndRecentQueryClientContext.RecentAndSavedFiltersQueryClientContext);
  const { logger } = (0, import_useFilterFieldLoggerContext.useFilterFieldLoggerContext)();
  const setCanPin = (0, import_useRecentFiltersStore.useRecentAndPinnedFiltersStore)(
    (state2) => state2.setCanPin,
    name
  );
  const state = (0, import_useRecentFiltersStore.useRecentAndPinnedFiltersStore)((state2) => {
    if (actionType === "save") {
      return state2.recentFilters;
    }
    return state2.pinnedFilters;
  }, name);
  const setState = (0, import_useRecentFiltersStore.useRecentAndPinnedFiltersStore)((state2) => {
    if (actionType === "save") {
      return state2.setRecentFilters;
    }
    return state2.setPinnedFilters;
  }, name);
  const action = (0, import_useRecentFiltersStore.useRecentAndPinnedFiltersStore)((state2) => {
    if (actionType === "save") {
      return state2.saveRecentFilter;
    }
    if (actionType === "pin") {
      return state2.pinFilter;
    }
    return state2.unpinFilter;
  }, name);
  const [previousState, setPreviousState] = (0, import_react.useState)(
    (0, import_lodash_es.cloneDeep)(state)
  );
  return (0, import_react_query.useMutation)(
    {
      mutationKey: [key],
      throwOnError: false,
      mutationFn: async (filterString) => {
        if (name === null) {
          return Promise.resolve();
        }
        if (filterString.trim() === "") {
          return Promise.resolve();
        }
        setPreviousState((0, import_lodash_es.cloneDeep)(state));
        const newState = action(filterString);
        await queryClient.cancelQueries({
          queryKey: [key]
        });
        try {
          await import_client_state.stateClient.setUserAppState({
            body: {
              validUntilTime: null,
              modificationInfo: {
                lastModifiedBy: currentUser.id,
                lastModifiedTime: /* @__PURE__ */ new Date()
              },
              value: JSON.stringify(newState)
            },
            key
          });
          return newState;
        } catch (error) {
          if ((0, import_client_state.isUnauthorized)(error) || (0, import_client_state.isForbidden)(error)) {
            logger?.warn(
              "FilterField pinned and recent error - permissions not met",
              error
            );
            setCanPin(false);
          } else {
            logger?.warn("FilterField pinned and recent error", error);
          }
          setState(previousState);
        }
        return state;
      },
      onError: (error) => {
        setState(previousState);
        queryClient.setQueryData([key], previousState);
        if ((0, import_client_state.isUnauthorized)(error) || (0, import_client_state.isForbidden)(error)) {
          logger?.warn(
            "FilterField pinned and recent error - permissions not met",
            error
          );
          setCanPin(false);
        }
      },
      onSuccess: (newState) => {
        if (newState !== void 0) {
          queryClient.setQueryData([key], newState);
        }
      }
    },
    queryClient
  );
}
