// @dynatrace/strato-components/filters/filter-field/hooks/recent-and-pinned/useRecentFilterSuggestions.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useRecentFilterSuggestions_exports = {};
__export(useRecentFilterSuggestions_exports, {
  useRecentFilterSuggestions: () => useRecentFilterSuggestions
});
module.exports = __toCommonJS(useRecentFilterSuggestions_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_useRecentFilters = require("./useRecentFilters.js");
var import_TimeIndicator = require("../../../../core/components/time-indicator/TimeIndicator.js");
var import_Text = require("../../../../typography/text/Text.js");
var import_editor_config = require("../../editor-state-configs/editor-config.js");
function useRecentFilterSuggestions(editorViewRef, filterNamespace = null) {
  const recentFilters = (0, import_useRecentFilters.useRecentFilters)(filterNamespace);
  const editorViewRefChange = (0, import_react.useRef)(editorViewRef.current);
  const [state, forceUpdate] = (0, import_react.useReducer)((x) => x + 1, 0);
  if (editorViewRefChange.current !== editorViewRef.current) {
    editorViewRefChange.current = editorViewRef.current;
    forceUpdate();
  }
  (0, import_react.useEffect)(() => {
    if (editorViewRef.current && recentFilters && recentFilters.length > 0) {
      const recentFiltersSuggestions = recentFilters.map(
        (storedFilter) => ({
          value: storedFilter.filter,
          insertionStrategy: "replace-all",
          icon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.HistoryIcon, {}),
          displayValue: storedFilter.filter,
          /* TODO
           * Discussed with @peter.affenzeller.
           * The details are by default rendered below the displayValue.
           * Only a container query moves it over to the right. Check with UX if this is
           */
          details: storedFilter.timeUsed && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { textStyle: "small", children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TimeIndicator.TimeIndicator, { time: storedFilter.timeUsed.getTime() }) })
        })
      );
      editorViewRef.current.dispatch({
        effects: [import_editor_config.recentFiltersEffect.of(recentFiltersSuggestions)]
      });
    }
  }, [state, editorViewRef.current, JSON.stringify(recentFilters)]);
}
