// @dynatrace/strato-components/esm/charts/tree-map/providers/TreeMapChartProviders.js
import { jsx } from "react/jsx-runtime";
import { useMemo } from "react";
import { StoreProvider } from "./StoreProvider.js";
import { ChartSeriesActionContext } from "../../core/components/series-action/ChartSeriesActionContext.js";
import { ColorRuleContext } from "../../core/context/ColorRuleContext.js";
import { ChartLegendContext } from "../context/ChartLegendContext.js";
import { ColorPaletteContext } from "../context/ColorPaletteContext.js";
import { FormatterContext } from "../context/FormatterContext.js";
import { LabelsDisplayContext } from "../context/LabelsDisplayContext.js";
import { MinMaxConfigContext } from "../context/MinMaxConfigContext.js";
import { NameAccessorContext } from "../context/NameAccessorContext.js";
import { ToolbarContext } from "../context/ToolbarContext.js";
import { TreeMapRawDataContext } from "../context/TreeMapRawDataContext.js";
import {
  DEFAULT_CUSTOM_TOOLTIP_CONFIG,
  TreeMapCustomTooltipContext
} from "../contexts/TreeMapCustomTooltipContext.js";
const TreeMapChartProviders = (props) => {
  const {
    data,
    seriesActions,
    formatter,
    truncationMode,
    legend,
    toolbar,
    labelsDisplay,
    nameAccessor,
    colorPalette,
    min: configMin,
    max: configMax,
    children,
    colorRules = [],
    unit,
    tooltip
  } = props;
  const min = configMin === "data-min" ? void 0 : configMin;
  const max = configMax === "data-max" ? void 0 : configMax;
  const memoizedRawData = useMemo(
    () => ({ ...data, unit: unit ?? data.unit }),
    [data, unit]
  );
  return /* @__PURE__ */ jsx(ColorRuleContext.Provider, { value: colorRules, children: /* @__PURE__ */ jsx(TreeMapRawDataContext.Provider, { value: memoizedRawData, children: /* @__PURE__ */ jsx(
    ChartSeriesActionContext.Provider,
    {
      value: {
        // eslint-disable-next-line @typescript-eslint/prefer-ts-expect-error, @typescript-eslint/ban-ts-comment
        // @ts-ignore Strict mode transition - needs to be investigated
        chartSeriesActions: seriesActions
      },
      children: /* @__PURE__ */ jsx(
        FormatterContext.Provider,
        {
          value: {
            formatter,
            truncationMode
          },
          children: /* @__PURE__ */ jsx(ColorPaletteContext.Provider, { value: colorPalette, children: /* @__PURE__ */ jsx(ChartLegendContext.Provider, { value: legend, children: /* @__PURE__ */ jsx(NameAccessorContext.Provider, { value: nameAccessor, children: /* @__PURE__ */ jsx(ToolbarContext.Provider, { value: toolbar, children: /* @__PURE__ */ jsx(LabelsDisplayContext.Provider, { value: labelsDisplay, children: /* @__PURE__ */ jsx(MinMaxConfigContext.Provider, { value: { min, max }, children: /* @__PURE__ */ jsx(
            TreeMapCustomTooltipContext.Provider,
            {
              value: tooltip ?? DEFAULT_CUSTOM_TOOLTIP_CONFIG,
              children: /* @__PURE__ */ jsx(StoreProvider, { children })
            }
          ) }) }) }) }) }) })
        }
      )
    }
  ) }) });
};
TreeMapChartProviders["displayName"] = "TreeMapChartProviders";
export {
  TreeMapChartProviders
};
