// @dynatrace/strato-components/esm/charts/core/components/tooltip/InteractiveChartTooltip/headers/TimebasedTooltipHeader.js
import { jsx, jsxs } from "react/jsx-runtime";
import { isBefore } from "date-fns";
import { formatDate } from "@dynatrace-sdk/units";
import { TimeResolution } from "./subheaders/TimeResolution.js";
import { FormattedTimeframe } from "../../utils/FormattedTimeframe.js";
import {
  StyledChartTooltipHeader,
  StyledFormattedTimeframe,
  StyledTitleContainer
} from "../components/InteractiveChartTooltip.styles.js";
import { SeriesBasedContent } from "../contents/SeriesBasedContent.js";
const TimebasedTooltipHeader = ({
  timeframe: [from, timeframeTo],
  item,
  seriesDisplayMode = "single-line"
}) => {
  const to = timeframeTo !== void 0 && isBefore(from, timeframeTo) ? timeframeTo : void 0;
  return /* @__PURE__ */ jsxs(StyledChartTooltipHeader, { children: [
    /* @__PURE__ */ jsxs(StyledTitleContainer, { gap: 8, children: [
      to && /* @__PURE__ */ jsx(TimeResolution, { from, to }),
      /* @__PURE__ */ jsx(StyledFormattedTimeframe, { children: to !== void 0 ? /* @__PURE__ */ jsx(
        FormattedTimeframe,
        {
          "data-testid": "formatted-timeframe",
          from,
          to
        }
      ) : formatDate(from, {
        day: "2-digit",
        month: "short",
        year: from.getFullYear() === (/* @__PURE__ */ new Date()).getFullYear() ? void 0 : "numeric",
        hour: "2-digit",
        minute: "2-digit",
        second: from.getSeconds() === 0 ? void 0 : "numeric"
      }) })
    ] }),
    item && /* @__PURE__ */ jsx(
      SeriesBasedContent,
      {
        item,
        highlighted: true,
        seriesDisplayMode
      }
    )
  ] });
};
export {
  TimebasedTooltipHeader
};
