// @dynatrace/strato-components/charts/core/components/annotations/components/overlays/AnnotationsOverlays.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var AnnotationsOverlays_exports = {};
__export(AnnotationsOverlays_exports, {
  AnnotationsOverlays: () => AnnotationsOverlays
});
module.exports = __toCommonJS(AnnotationsOverlays_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_AnnotationsOverlays_css = require("./AnnotationsOverlays.sty.js");
var import_constants = require("../../constants.js");
var import_AnnotationsConfig_context = require("../../contexts/AnnotationsConfig.context.js");
var import_useOverlayEvents = require("../../hooks/useOverlayEvents.js");
const AnnotationsOverlays = ({ height }) => {
  const { labelsWidth, trackWidth } = (0, import_react.useContext)(import_AnnotationsConfig_context.AnnotationsConfigContext);
  const { handleMouseMove, handleMouseLeave, handleClick } = (0, import_useOverlayEvents.useOverlayEvents)();
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      onMouseLeave: handleMouseLeave,
      onMouseMove: handleMouseMove,
      onClick: handleClick,
      className: import_AnnotationsOverlays_css.annotationsOverlay,
      style: {
        height,
        width: trackWidth,
        left: labelsWidth,
        top: import_constants.MARGIN_TOP_STROKE
      }
    }
  );
};
