// @dynatrace/strato-components/esm/charts/xy-chart/components/overlay/XYChartOverlay.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { every } from "lodash-es";
import { useEffect, useRef } from "react";
import { XYChartSynchronizedCrosshair } from "./XYChartSynchronizedCrosshair.js";
import { useGridPlotSize } from "../../../core/components/chart-grid/hooks/useGridPlotSize.js";
import { InteractionsOverlay } from "../../../core/components/interactions-overlay/InteractionsOverlay.js";
import { useOverlay } from "../../../core/components/overlay-tooltip/contexts/OverlayContext.js";
import {
  useInteractionModeState,
  useSelectState,
  useSetState,
  useZoomState
} from "../../store/selectors.js";
import { XYChartSelectionArea } from "../selection/XYChartSelectionArea.js";
import { XYChartToolbar } from "../toolbar/XYChartToolbar.js";
import { XYChartOverlayTooltip } from "../tooltip/XYChartOverlayTooltip.js";
import { XYChartCrosshair } from "./explore/XYChartCrosshair.js";
import { XYChartLinearPoint } from "./explore/XYChartLinearPoint.js";
import { XYChartRectanglePoint } from "./explore/XYChartRectanglePoint.js";
import { useOverlayEvents } from "./hooks/useOverlayEvents.js";
import { isOwnEvent } from "../../../core/components/interactions-overlay/utils/is-own-event.js";
import {
  useSelector,
  useTooltipSnapshot
} from "../../../core/components/overlay-tooltip/selectors.js";
import { SynchronizedSelectionArea } from "../../../core/components/shared-chart-interactions/components/SynchronizedSelectionArea.js";
import { useSynchronizedSelection } from "../../../core/components/shared-chart-interactions/hooks/useSynchronizedSelection.js";
import { useToolbarHandlerRef } from "../../../core/components/toolbar/hooks/useToolbarHandlerRef.js";
import { zIndicesXYChartAreas } from "../../constants/xy-chart-z-index-areas.js";
import { XYChartSelectionAreaRefContext } from "../../context/XYChartSelectionAreaRef.context.js";
import { useSyncedXScaleByPriority } from "../../hooks/useSyncedXScaleByPriority.js";
import { useXYChartData } from "../../hooks/useXYChartData.js";
import { useXYChartId } from "../../hooks/useXYChartId.js";
const XYChartOverlay = () => {
  const setState = useSetState();
  const handlerRef = useToolbarHandlerRef();
  const overlay = useOverlay();
  const selectionRef = useRef(null);
  const { width, height } = useGridPlotSize();
  useSelector(({ inBounds: inBounds2, pinned: pinned2 }) => `${inBounds2}-${pinned2}`);
  const { inBounds, pinned } = useTooltipSnapshot();
  const {
    handleModeChange,
    handleHover,
    handleTogglePin,
    handleReset,
    handleLeave,
    handlePanChange,
    handleSelectChange,
    handleSelectEnd,
    handleZoomChange,
    handleSelectStart,
    resetTooltip
  } = useOverlayEvents();
  const mode = useInteractionModeState();
  const selectState = useSelectState();
  const synchronizedSelection = useSynchronizedSelection();
  const chartId = useXYChartId();
  const syncedXScale = useSyncedXScaleByPriority();
  const { axesMetadata } = useXYChartData();
  const { zoomOutEnabled } = useZoomState();
  const isZoomDisabled = !axesMetadata.isZoomEnabled;
  const isPanDisabled = !(axesMetadata.isPanEnabled && zoomOutEnabled);
  useEffect(() => {
    if (every([
      selectState,
      !synchronizedSelection.selectedChartId,
      !synchronizedSelection?.disabled
    ])) {
      setState((s) => ({ ...s, select: void 0 }));
    }
  }, [synchronizedSelection.selectedChartId]);
  useEffect(() => {
    if (synchronizedSelection.selectedChartId === chartId && !selectState) {
      synchronizedSelection.resetSelectionState?.();
    }
  }, [!!selectState]);
  const showHighlightedData = selectState === void 0 || selectState?.variant === "select";
  const handleFocus = (event) => {
    if (!event || event.cancelable) {
      return;
    }
    if (isOwnEvent(event)) {
      handlerRef.current.handleFocus();
    }
  };
  return /* @__PURE__ */ jsxs(
    InteractionsOverlay,
    {
      style: {
        zIndex: zIndicesXYChartAreas.overlays,
        userSelect: selectState?.finished === false ? "none" : "auto"
      },
      ref: overlay.overlayRef,
      mode,
      pinned,
      selectionArea: selectState ? [selectState?.start, selectState?.end] : void 0,
      width,
      height,
      onModeChange: handleModeChange,
      onReset: handleReset,
      onLeave: handleLeave,
      onPanChange: handlePanChange,
      onHover: handleHover,
      onFocus: handleFocus,
      onTogglePin: ({ pin, position }) => handleTogglePin(pin, position),
      onSelectStart: handleSelectStart,
      onSelectChange: handleSelectChange,
      onSelectEnd: (event) => handleSelectEnd(event, selectionRef),
      onZoomChange: handleZoomChange,
      isZoomDisabled,
      isZoomXDisabled: isZoomDisabled,
      isPanDisabled,
      children: [
        (inBounds || pinned || selectState) && /* @__PURE__ */ jsxs(Fragment, { children: [
          /* @__PURE__ */ jsx(XYChartOverlayTooltip, {}),
          /* @__PURE__ */ jsx(XYChartCrosshair, { chartRef: overlay.overlayRef }),
          showHighlightedData && /* @__PURE__ */ jsxs(Fragment, { children: [
            /* @__PURE__ */ jsx(XYChartRectanglePoint, {}),
            /* @__PURE__ */ jsx(XYChartLinearPoint, {})
          ] })
        ] }),
        /* @__PURE__ */ jsx(XYChartSynchronizedCrosshair, {}),
        /* @__PURE__ */ jsxs(XYChartSelectionAreaRefContext.Provider, { value: selectionRef, children: [
          /* @__PURE__ */ jsx(XYChartToolbar, { onMouseEnter: resetTooltip, handlerRef }),
          /* @__PURE__ */ jsx(XYChartSelectionArea, {})
        ] }),
        synchronizedSelection.selectedChartId && synchronizedSelection.selectedChartId !== chartId && syncedXScale && /* @__PURE__ */ jsx(SynchronizedSelectionArea, { scale: syncedXScale })
      ]
    }
  );
};
XYChartOverlay["displayName"] = "XYChartOverlay";
export {
  XYChartOverlay
};
