// @dynatrace/strato-components/esm/forms/number-input-v2/NumberInput.js
import { jsx } from "react/jsx-runtime";
import {
  forwardRef,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState
} from "react";
import { useIntl } from "react-intl";
import { applyStep } from "./utils/calculations.js";
import { checkFormat, parseNumber } from "./utils/format.js";
import { addError, getErrorKey } from "./utils/validation.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useId } from "../../core/hooks/useId.js";
import { useMergeRefs } from "../../core/hooks/useMergeRefs.js";
import { mergeProps } from "../../core/utils/merge-props.js";
import { BaseInput } from "../_base-input/BaseInput.js";
import { BaseInputButton } from "../_base-input/BaseInputButton.js";
import {
  BaseInputPrefix,
  BaseInputSuffix
} from "../_base-input/BaseInputContent.js";
import { useFormFieldMessagesContext } from "../forms-core/hooks/useFormFieldMessagesContext.js";
import { FormFieldMessagesAction } from "../forms-core/state/formfield-messages-state-reducer.js";
import { getSafeRegionalFormat } from "../forms-core/time/helpers.js";
const COMPONENT_NAME = "NumberInput";
const VALID_PATTERN = "^[+\\-]?\\d+(?:[\\.,]\\d+)?(?:[eE][+\\-]?\\d+)?$";
const VALID_REGEXP = new RegExp(VALID_PATTERN);
const FORMATTABLE_PATTERN = /^[-+]?[\d.,'\u2019 ٫٬\u00A0\u1680\u2000-\u200A\u202F\u205F\u3000]+(?:[eE][-+]?\d+)?$/u;
const INTERMEDIATE_VALID_PATTERN = /^[+-]?\d*(?:[.,]\d*)?(?:[eE][+-]?\d*)?$/;
const NumberInputRoot = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const intl = useIntl();
  const defaultPlaceholder = intl.formatMessage({
    defaultMessage: "Add number",
    id: "rrkUQoVNb46Jbd/8",
    description: "Default placeholder for the number input."
  });
  const {
    onChange,
    onKeyDown,
    value: propValue,
    defaultValue,
    required,
    readOnly = false,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "aria-disabled": ariaDisabledProp,
    min,
    max,
    step,
    stepMultiplier = 1,
    placeholder = defaultPlaceholder,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
  const { state, dispatch } = useFormFieldMessagesContext();
  const errorId = useId();
  const isPasted = useRef(false);
  const isInternalChanged = useRef(defaultValue !== void 0);
  const internalRef = useRef(null);
  const inputRef = useMergeRefs([forwardedRef, internalRef]);
  const [internalValue, setInternalValue] = useState(
    defaultValue ?? propValue ?? null
  );
  const setValueNative = useCallback(
    (newValue, triggerChange = true) => {
      const input = internalRef.current?.inputRef;
      if (input) {
        let stringValue;
        if (newValue === null) {
          stringValue = "";
        } else if (typeof newValue === "number") {
          stringValue = new Intl.NumberFormat(getSafeRegionalFormat(), {
            useGrouping: false,
            maximumFractionDigits: 17
          }).format(newValue);
        } else {
          stringValue = String(newValue);
        }
        const desc = Object.getOwnPropertyDescriptor(
          HTMLInputElement.prototype,
          "value"
        );
        desc?.set?.call(input, stringValue);
        input.value = stringValue;
        if (triggerChange) {
          const parsedValue = checkFormat(stringValue);
          isInternalChanged.current = true;
          setInternalValue(parsedValue || stringValue);
          if (VALID_REGEXP.exec(parsedValue)) {
            onChange?.(parseNumber(parsedValue));
          } else if (parsedValue === "") {
            onChange?.(null);
          }
        }
      }
    },
    [onChange]
  );
  const value = useMemo(() => {
    const isChanged = isInternalChanged.current;
    isInternalChanged.current = false;
    if (isChanged) {
      return internalValue;
    }
    if (!propValue && propValue !== 0) {
      return "";
    }
    const formattedPropValue = new Intl.NumberFormat(getSafeRegionalFormat(), {
      useGrouping: false,
      maximumFractionDigits: 17
      // Max precision for a JS number
    }).format(propValue);
    return formattedPropValue;
  }, [propValue, internalValue]);
  const handleChange = useCallback(
    (e) => {
      const parsedValue = checkFormat(e.target.value);
      if (VALID_REGEXP.exec(parsedValue)) {
        onChange?.(parseNumber(parsedValue));
      }
      if (parsedValue === "") {
        onChange?.(null);
      }
      if (isPasted.current && FORMATTABLE_PATTERN.exec(parsedValue)) {
        isInternalChanged.current = true;
        setInternalValue(parsedValue);
        setValueNative(parsedValue, false);
      } else if (INTERMEDIATE_VALID_PATTERN.exec(e.target.value)) {
        isInternalChanged.current = true;
        setInternalValue(parsedValue);
      } else {
        setValueNative(value ?? "", false);
      }
      isPasted.current = false;
    },
    [onChange, setInternalValue, value, setValueNative]
  );
  const handleStep = useCallback(
    (direction) => {
      setValueNative(
        applyStep(value, step, stepMultiplier, min, max, direction)
      );
    },
    [value, stepMultiplier, step, min, max, setValueNative]
  );
  const handleKeyDown = useCallback(
    (e) => {
      if (readOnly) {
        onKeyDown?.(e);
        return;
      }
      if (e.key === "ArrowUp") {
        e.preventDefault();
        handleStep(1);
      }
      if (e.key === "ArrowDown") {
        e.preventDefault();
        handleStep(-1);
      }
      if (e.key === "Home" && min !== void 0) {
        e.preventDefault();
        setValueNative(min);
      }
      if (e.key === "End" && max !== void 0) {
        e.preventDefault();
        setValueNative(max);
      }
      onKeyDown?.(e);
    },
    [handleStep, onKeyDown, setValueNative, min, max, readOnly]
  );
  useEffect(() => {
    const el = internalRef?.current;
    if (!el) {
      return;
    }
    const onWheel = (e) => {
      if (document.activeElement !== e.target) {
        return;
      }
      e.preventDefault();
      handleStep(e.deltaY < 0 ? 1 : -1);
    };
    el.element?.addEventListener("wheel", onWheel, { passive: false });
    return () => el.element?.removeEventListener("wheel", onWheel);
  }, [handleStep]);
  useEffect(() => {
    if (!value && value !== 0) {
      return;
    }
    const errorMessage = getErrorKey(value, min, max, step);
    if (errorMessage) {
      addError(errorMessage, errorId, internalRef.current?.inputRef, dispatch);
      return;
    }
    dispatch?.({
      type: FormFieldMessagesAction.REMOVEMESSAGE,
      payload: {
        id: errorId,
        messageType: "formcontrol"
      }
    });
  }, [
    value,
    min,
    max,
    step,
    state.isTouched,
    state.isValid,
    dispatch,
    errorId
  ]);
  return /* @__PURE__ */ jsx(
    BaseInput,
    {
      ref: inputRef,
      ...mergeProps(restProps, behavioralTrackingProps),
      type: "text",
      inputMode: "decimal",
      autoComplete: "off",
      value: value ?? "",
      required,
      readOnly,
      onChange: handleChange,
      onKeyDown: handleKeyDown,
      onPaste: () => isPasted.current = true,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "aria-disabled": ariaDisabledProp,
      "data-dt-component": COMPONENT_NAME,
      min,
      max,
      step,
      pattern: VALID_PATTERN,
      placeholder
    }
  );
});
const NumberInput = Object.assign(NumberInputRoot, {
  Prefix: BaseInputPrefix,
  Suffix: BaseInputSuffix,
  Button: BaseInputButton
});
NumberInput.displayName = COMPONENT_NAME;
export {
  NumberInput
};
