// @dynatrace/strato-components/esm/charts/timeseries/components/overlays/utils/tooltip/find-closest-points.js
import { isNaN } from "lodash-es";
import { OVERLAY_PADDING } from "../../../../../core/components/data-point/constants.js";
import { safeGet } from "../../../../../core/utils/safe-get.js";
import { sanitizeZeroValues } from "../../../../../core/utils/sanitize-zero-values.js";
import { getClusterBoundariesAtPosition } from "../../../../utils/binary-search.js";
import { getDatapointTime } from "../../../../utils/timeseries.js";
import { getYScaleByUnit } from "../../../../utils/yAxis.js";
const getClosestPoint = (xPoints, mouseY = 0) => {
  let aux = Infinity;
  let auxPoint = xPoints[0];
  xPoints.forEach((elem) => {
    if (Math.abs(elem.y - mouseY) < aux) {
      aux = Math.abs(elem.y - mouseY);
      auxPoint = elem;
    }
  });
  return auxPoint;
};
function xAccessor(xAxisScale) {
  return (datapoint) => xAxisScale(getDatapointTime(datapoint.start, datapoint.end));
}
function roundTo2Decimals(num) {
  const factor = 100;
  return Math.round(num * factor) / factor;
}
function clusterAccessor(xAxisScale, mouseX) {
  return (datapoint) => roundTo2Decimals(
    Math.abs(
      xAxisScale(getDatapointTime(datapoint.start, datapoint.end)) - mouseX
    )
  );
}
function mapToClosestPoints(datapointsMap, xAxisScale, yAxis) {
  return (datapoint) => {
    const datapointMap = safeGet(datapointsMap, datapoint);
    const {
      isGap,
      color,
      value,
      unit,
      y0,
      y1,
      cumulativeValue,
      seriesId,
      start,
      end,
      center,
      variant
    } = datapointMap;
    if (isGap) {
      return [];
    }
    const yScale = getYScaleByUnit(unit, yAxis);
    const x = xAxisScale(getDatapointTime(start, end, center));
    if (variant === "band") {
      const pointY0 = sanitizeZeroValues(yScale, y0 ?? 0);
      const pointY1 = sanitizeZeroValues(yScale, y1 ?? 0);
      return [
        { id: seriesId, x, y: pointY0, color, datapoint },
        { id: seriesId, x, y: pointY1, color, datapoint }
      ];
    }
    const y = sanitizeZeroValues(
      yScale,
      cumulativeValue ? cumulativeValue : value ?? 0
    );
    if (!isNaN(y)) {
      return [{ id: seriesId, x, y, color, datapoint }];
    }
    return [];
  };
}
const findClosestPoints = (mouseX, datapoints, xAxisScale, yAxis, datapointsMap) => {
  const [leftIndex, rightIndex] = getClusterBoundariesAtPosition(
    mouseX,
    datapoints,
    xAccessor(xAxisScale),
    clusterAccessor(xAxisScale, mouseX)
  );
  const toClosestPoint = mapToClosestPoints(datapointsMap, xAxisScale, yAxis);
  const closestPoints = [];
  for (let i = leftIndex; i <= rightIndex; i++) {
    closestPoints.push(...toClosestPoint(datapoints[i]));
  }
  return closestPoints;
};
const filterOutOfBounds = (overlaysWidth, overlaysHeight) => (datapoint) => {
  return datapoint.y >= OVERLAY_PADDING && datapoint.y <= overlaysHeight + OVERLAY_PADDING && datapoint.x >= OVERLAY_PADDING && datapoint.x <= overlaysWidth + OVERLAY_PADDING;
};
export {
  filterOutOfBounds,
  findClosestPoints,
  getClosestPoint
};
