// @dynatrace/strato-components/esm/charts/timeseries/components/overlays/utils/tooltip/timeseries-tooltip.js
import { isUndefined } from "lodash-es";
import { defaultUnitFormatter } from "../../../../../core/utils/defaultUnitFormatter.js";
import { safeGet } from "../../../../../core/utils/safe-get.js";
const buildBandDataPointValue = (y0, y1, unit, formatter, originalUnit, absoluteY0, absoluteY1) => {
  const [min, max] = [y0, y1].sort((a, b) => a - b);
  const result = [`${formatter(min, unit)} \u2013 ${formatter(max, unit)}`];
  if (absoluteY0 !== void 0 && absoluteY1 !== void 0) {
    const [absoluteMin, absoluteMax] = [absoluteY0, absoluteY1].sort(
      (a, b) => a - b
    );
    result.unshift(
      `${formatter(absoluteMin, originalUnit)} \u2013 ${formatter(
        absoluteMax,
        originalUnit
      )}`
    );
  }
  return result;
};
const buildValuesBasedOnValueRepresentation = (value, unit, originalUnit, formatter, absoluteValue) => {
  const isRelativeRepresentation = absoluteValue !== void 0;
  if (isRelativeRepresentation) {
    return [
      defaultUnitFormatter(originalUnit)(absoluteValue),
      formatter(value)
    ];
  }
  return [formatter(value, unit)];
};
const tooltipItemsMemoized = /* @__PURE__ */ new Map();
const toTooltipItems = (xPoints, formatters, datapointsMap, units) => {
  const tooltipItems = [];
  const bandSeries = {};
  const [leftUnit] = units;
  const [leftFormatter, rightFormatter] = formatters;
  xPoints.forEach((xPoint, index) => {
    const xPointId = `${xPoint.id}-${index}-${xPoint.y}-${xPoint.x}`;
    let tooltipItem;
    const memoedTooltipItem = tooltipItemsMemoized.get(xPointId);
    if (memoedTooltipItem) {
      tooltipItem = memoedTooltipItem;
    } else {
      const datapointMetadata = safeGet(datapointsMap, xPoint.datapoint);
      const {
        seriesId,
        value,
        absoluteValue,
        y0,
        absoluteY0,
        y1,
        absoluteY1,
        color,
        unit,
        originalUnit,
        seriesName,
        variant
      } = datapointMetadata;
      const formatter = unit === leftUnit ? leftFormatter : rightFormatter;
      const baseTooltipItem = {
        id: `${seriesId}-${index}`,
        name: seriesName,
        color,
        datapoint: xPoint.datapoint
      };
      if (variant === "band") {
        if (!bandSeries[xPoint.id]) {
          tooltipItems.push({
            ...baseTooltipItem,
            asyncValues: () => buildBandDataPointValue(
              // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
              y0,
              // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
              y1,
              unit,
              formatter,
              originalUnit,
              absoluteY0,
              absoluteY1
            )
          });
          bandSeries[xPoint.id] = true;
          return;
        }
        bandSeries[xPoint.id] = false;
        return;
      }
      tooltipItem = {
        ...baseTooltipItem,
        asyncValues: () => buildValuesBasedOnValueRepresentation(
          value ?? 0,
          unit,
          originalUnit ? originalUnit : "",
          formatter,
          absoluteValue
        )
      };
      tooltipItemsMemoized.set(xPointId, tooltipItem);
    }
    tooltipItems.push(tooltipItem);
  });
  return tooltipItems;
};
function filterGroupedPoints(closestPoint, points, groupedPoints) {
  const closestPointGroupIds = groupedPoints.filter((group) => group.map((g) => g.id).includes(closestPoint.id)).flatMap((group) => group.map((point) => point.id));
  return points.filter((point) => closestPointGroupIds.includes(point.id));
}
function filterTimestampsAndTimeframes(points, closestPoint) {
  const filterTimestamps = (point) => isUndefined(point.datapoint.end);
  const filterTimeframes = (point) => !isUndefined(point.datapoint.end);
  const hasTimestampClosestPoint = isUndefined(closestPoint.datapoint.end);
  return hasTimestampClosestPoint ? points.filter(filterTimestamps) : points.filter(filterTimeframes);
}
export {
  filterGroupedPoints,
  filterTimestampsAndTimeframes,
  toTooltipItems
};
