// @dynatrace/strato-components/esm/tables/DataTable/features/RowDetails/RowDetails.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import { uniqueId } from "lodash-es";
import {
  useMemo,
  Children
} from "react";
import { useIntl } from "react-intl";
import { isElement } from "react-is";
import Borders from "@dynatrace/strato-design-tokens/borders";
import Spacings from "@dynatrace/strato-design-tokens/spacings";
import {
  ChevronDownSmallIcon,
  ChevronRightSmallIcon
} from "@dynatrace/strato-icons";
import { getRowDetailsRowModel } from "./details-row-model.js";
import { rowDetailsWrapperCSS } from "./RowDetails.sty.js";
import { Button } from "../../../../buttons/button/Button.js";
import { useControllableState } from "../../../../core/hooks/useControllableState.js";
import { DataTableDefaultCell } from "../../components/DataTableDefaultCell.js";
import { DATA_TABLE_ROW_DETAILS_COLUMN_ID } from "../../constants.js";
import { useDataTable } from "../../contexts/DataTableContext.js";
import { assignRowMethods } from "../../hooks/useTable/utils.js";
import { dataTableMessages } from "../../messages.js";
function DataTableExpandableRowTemplate(props) {
  return null;
}
function DataTableExpandableRowWrapper(props) {
  const {
    className: consumerClassName,
    "data-testid": dataTestId,
    style: consumerStyle,
    ...remainingProps
  } = props;
  const table = useDataTable();
  return /* @__PURE__ */ jsx(
    "div",
    {
      "data-testid": dataTestId,
      ...remainingProps,
      className: clsx(
        consumerClassName,
        rowDetailsWrapperCSS({
          rowDensity: table.getVisualVariantOptions().rowDensity
        })
      ),
      style: consumerStyle
    }
  );
}
function isDataTableExpandableRowTemplate(child) {
  return isElement(child) && child.type === DataTableExpandableRowTemplate;
}
const DataTableRowDetails = {
  getInitialState(state) {
    return {
      rowDetails: {},
      ...state
    };
  },
  createTable(table) {
    table.rowDetailsTableId = uniqueId("DataTable-");
    table.getPreRowDetailsRowModel = () => table.getPaginationRowModel();
    table.getRowDetailsRowModel = () => {
      if (!table._getRowDetailsRowModel && table.options.getRowDetailsRowModel) {
        table._getRowDetailsRowModel = table.options.getRowDetailsRowModel(table);
      }
      if (!table._getRowDetailsRowModel) {
        return table.getPreRowDetailsRowModel();
      }
      return table._getRowDetailsRowModel();
    };
    assignRowMethods(table, {
      /**
       * Determines if the details row for a particular row is open.
       */
      getIsDetailsOpen() {
        const detailOpenState = table.getState().rowDetails ?? false;
        if (typeof detailOpenState === "boolean") {
          return detailOpenState;
        }
        return detailOpenState[this.id] === true;
      },
      /**
       * Defines the grid position for the row details row.
       */
      getRowDetailsGridPosition() {
        return {
          gridColumnStart: 1,
          gridColumnEnd: `span ${table.getVisibleLeafColumns().length}`
        };
      },
      /**
       * Returns the aria control id to connect the expanding button to the
       * row details wrapper.
       */
      getRowDetailsAriaControlsId() {
        return `${table.rowDetailsTableId}-${this.id}-expand`;
      },
      /**
       * Determines if the details for this row can be shown.
       */
      getCanOpenDetails() {
        if (typeof table.options.rowDetailsDisableRow === "function") {
          return !table.options.rowDetailsDisableRow(this.original);
        }
        return true;
      },
      /**
       * Control command to toggle the state of the rowDetails.
       */
      toggleDetails() {
        const scrollOffset = table.getRowScrollOffset();
        table.options.onRowDetailsOpenChange?.(
          (prev) => {
            let newRowDetailsState = {};
            if (typeof prev === "boolean" && prev === true) {
              const flatRows = table.getRowModel().flatRows;
              for (const flatRow of flatRows) {
                if (this.getCanOpenDetails() && this.id !== flatRow.id) {
                  newRowDetailsState[flatRow.id] = true;
                }
              }
              return newRowDetailsState;
            }
            newRowDetailsState = { ...prev };
            if (newRowDetailsState[this.id] === true) {
              delete newRowDetailsState[this.id];
            } else {
              newRowDetailsState[this.id] = true;
            }
            return newRowDetailsState;
          }
        );
        requestAnimationFrame(() => {
          table.scrollToRowOffset(scrollOffset);
        });
      }
    });
  }
};
function isRowDetailsColumnDef(column) {
  return column.id === ROW_DETAILS_COLUMN.id;
}
function RowDetailsCell(props) {
  const { row } = props;
  const intl = useIntl();
  const isOpen = row.getIsDetailsOpen();
  const Icon = isOpen ? /* @__PURE__ */ jsx(ChevronDownSmallIcon, {}) : /* @__PURE__ */ jsx(ChevronRightSmallIcon, {});
  return /* @__PURE__ */ jsx(DataTableDefaultCell, { children: /* @__PURE__ */ jsx(
    Button,
    {
      size: "condensed",
      variant: "default",
      style: {
        borderRadius: Borders.Radius.Field.Subdued,
        paddingLeft: Spacings.Size2,
        paddingRight: Spacings.Size2,
        paddingTop: 0,
        paddingBottom: 0
      },
      "aria-label": intl.formatMessage(dataTableMessages.rowDetailsTrigger, {
        isDetailsOpen: isOpen,
        id: row.id
      }),
      "aria-expanded": isOpen,
      "aria-controls": isOpen ? row.getRowDetailsAriaControlsId() : void 0,
      onClick: () => row.toggleDetails(),
      disabled: !row.getCanOpenDetails(),
      children: /* @__PURE__ */ jsx(Button.Prefix, { children: Icon })
    }
  ) });
}
const ROW_DETAILS_COLUMN = {
  id: DATA_TABLE_ROW_DETAILS_COLUMN_ID,
  alignment: "left",
  accessorFn: () => "",
  header: () => {
    return "";
  },
  width: "content",
  enableSorting: false,
  enableResizing: false,
  enableUserActions: false,
  enableHiding: false,
  meta: {
    builtin: true,
    builtinPlacement: "prefix",
    hasCustomCellRenderer: true,
    hasCustomHeaderRenderer: true,
    cellDeps: (cell) => {
      return [cell.row.getIsDetailsOpen()];
    }
  },
  cell: ({ row }) => /* @__PURE__ */ jsx(RowDetailsCell, { row })
};
function useRowDetails(props, options) {
  const { children } = props;
  const rowDetailsTemplateProps = useMemo(() => {
    for (const child of Children.toArray(children)) {
      if (isDataTableExpandableRowTemplate(child)) {
        return child?.props;
      }
    }
  }, [children]);
  const [openRowDetails, setOpenRowDetails] = useControllableState({
    value: rowDetailsTemplateProps?.expandedRows,
    defaultValue: rowDetailsTemplateProps?.defaultExpandedRows,
    onChange: (value) => {
      if (value !== true) {
        rowDetailsTemplateProps?.onExpandedRowsChange?.(value);
      }
    }
  });
  const template = rowDetailsTemplateProps?.children;
  if (template !== void 0 && typeof template === "function") {
    options.enableRowDetails = true;
    options.rowDetailsTemplate = template;
    options.getRowDetailsRowModel = getRowDetailsRowModel();
    options.rowDetailsDisableRow = rowDetailsTemplateProps?.disableExpand;
    options.state = {
      ...options.state,
      rowDetails: openRowDetails
    };
    options.onRowDetailsOpenChange = setOpenRowDetails;
  }
}
export {
  DataTableExpandableRowTemplate,
  DataTableExpandableRowWrapper,
  DataTableRowDetails,
  ROW_DETAILS_COLUMN,
  isRowDetailsColumnDef,
  useRowDetails
};
