// @dynatrace/strato-components/esm/tables/DataTable/features/RowDetails/RowDetailsElement.js
import { jsx } from "react/jsx-runtime";
import { rowDetailsRow } from "./RowDetails.sty.js";
import { getDataTableBorderWidth } from "../../../utils/data-table-border-width.js";
import { useDataTable } from "../../contexts/DataTableContext.js";
function RowDetailsElement(props) {
  const {
    originalRow,
    virtualRow,
    virtualItemIndex,
    rowSeparation,
    rowParity,
    rowMeasure,
    isLastRowOnPage: isLastRow,
    isDraggedRow = false,
    children
  } = props;
  const table = useDataTable();
  let width = table.getRowVirtualizationContainer()?.offsetWidth;
  if (width && table.getVisualVariantOptions().contained) {
    width -= 2 * getDataTableBorderWidth();
  }
  return /* @__PURE__ */ jsx(
    "div",
    {
      role: "row",
      id: originalRow.getRowDetailsAriaControlsId(),
      style: {
        ...originalRow.getGridRowPosition(virtualItemIndex),
        ...originalRow.getRowDetailsGridPosition(),
        maxWidth: width,
        position: "sticky",
        left: 0
      },
      className: rowDetailsRow({
        rowSeparation,
        rowParity,
        isDraggedRow,
        isLastRow
      }),
      ref: rowMeasure,
      "data-rowid": originalRow.id,
      "data-row": rowMeasure ? virtualRow.index : void 0,
      "data-column": 0,
      children
    }
  );
}
export {
  RowDetailsElement
};
