// @dynatrace/strato-components/charts/core/components/chart-layout/utils/get-areas-size-on-drag.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var get_areas_size_on_drag_exports = {};
__export(get_areas_size_on_drag_exports, {
  getAreasSizeOnDragHandlers: () => getAreasSizeOnDragHandlers,
  getDefaultAreasSize: () => getDefaultAreasSize
});
module.exports = __toCommonJS(get_areas_size_on_drag_exports);
var import_calculate_graph_size = require("./calculate-graph-size.js");
var import_constants = require("../../../constants.js");
var import_constants2 = require("../types/constants.js");
const getAreasWidthSize = (containerHeight, containerWidth, minWidth, maxWidth, resizerWrapperSize, draggedAmount, legendContainer) => {
  const minLegendWidth = Math.min(containerWidth * minWidth, import_constants2.MIN_LEGEND_WIDTH);
  if (legendContainer.clientWidth - draggedAmount <= minLegendWidth) {
    const legend = {
      height: containerHeight,
      width: minLegendWidth
    };
    const graph = {
      width: (0, import_calculate_graph_size.calculateGraphSize)(
        containerWidth,
        minLegendWidth,
        resizerWrapperSize
      ),
      height: containerHeight
    };
    return { legend, graph };
  }
  const availableWidth = containerWidth - resizerWrapperSize;
  const maxWidthAllowed = Math.max(
    availableWidth * maxWidth,
    availableWidth - import_constants2.MIN_PLOT_SIZE
  );
  if (legendContainer.clientWidth - draggedAmount >= maxWidthAllowed) {
    const legend = {
      height: containerHeight,
      width: maxWidthAllowed
    };
    const graph = {
      width: (0, import_calculate_graph_size.calculateGraphSize)(
        containerWidth,
        maxWidthAllowed,
        resizerWrapperSize
      ),
      height: containerHeight
    };
    return { legend, graph };
  }
};
const getAreasHeightSize = (containerHeight, containerWidth, minHeight, maxHeight, resizerWrapperSize, draggedAmount, legendContainer) => {
  const minLegendHeight = Math.min(
    containerHeight * minHeight,
    import_constants2.MIN_LEGEND_HEIGHT
  );
  if (legendContainer.clientHeight - draggedAmount <= minLegendHeight) {
    const legend = {
      height: minLegendHeight,
      width: containerWidth
    };
    const graph = {
      height: (0, import_calculate_graph_size.calculateGraphSize)(
        containerHeight,
        minLegendHeight,
        resizerWrapperSize
      ),
      width: containerWidth
    };
    return { legend, graph };
  }
  const availableHeight = containerHeight - resizerWrapperSize;
  const maxHeightAllowed = Math.max(
    availableHeight * maxHeight,
    availableHeight - import_constants2.MIN_PLOT_SIZE
  );
  if (legendContainer.clientHeight - draggedAmount >= maxHeightAllowed) {
    const legend = {
      height: maxHeightAllowed,
      width: containerWidth
    };
    const graph = {
      height: (0, import_calculate_graph_size.calculateGraphSize)(
        containerHeight,
        maxHeightAllowed,
        resizerWrapperSize
      ),
      width: containerWidth
    };
    return { legend, graph };
  }
};
const getDefaultAreasSize = (areasSizes, draggedAmount, containerSize, resizerWrapperSize, orientation) => {
  const { graph, legend } = areasSizes;
  const isVertical = orientation === import_constants.ORIENTATION.Vertical;
  const legendHeight = isVertical ? legend.height - draggedAmount : legend.height;
  const legendWidth = isVertical ? legend.width : legend.width - draggedAmount;
  const graphHeight = isVertical ? (0, import_calculate_graph_size.calculateGraphSize)(containerSize, legendHeight, resizerWrapperSize) : graph.height;
  const graphWidth = isVertical ? graph.width : (0, import_calculate_graph_size.calculateGraphSize)(containerSize, legendWidth, resizerWrapperSize);
  return {
    legend: { height: legendHeight, width: legendWidth },
    graph: { height: graphHeight, width: graphWidth }
  };
};
const getAreasSizeOnDragHandlers = (containerHeight, containerWidth, legendContainer, resizerWrapperSize) => ({
  getAreasHeightSizeOnDrag: (minHeight, maxHeight, verticalDraggedAmount = 0) => getAreasHeightSize(
    containerHeight,
    containerWidth,
    minHeight,
    maxHeight,
    resizerWrapperSize,
    verticalDraggedAmount,
    legendContainer
  ),
  getAreasWidthSizeOnDrag: (minWidth, maxWidth, horizontalDraggedAmount = 0) => getAreasWidthSize(
    containerHeight,
    containerWidth,
    minWidth,
    maxWidth,
    resizerWrapperSize,
    horizontalDraggedAmount,
    legendContainer
  )
});
