// @dynatrace/strato-components/charts/core/components/chart-layout/utils/get-chart-layout-settings.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var get_chart_layout_settings_exports = {};
__export(get_chart_layout_settings_exports, {
  getChartLayoutSettings: () => getChartLayoutSettings
});
module.exports = __toCommonJS(get_chart_layout_settings_exports);
var import_lodash_es = require("lodash");
var import_extract_layout_slots = require("./extract-layout-slots.js");
var import_orientation = require("./orientation.js");
var import_sanitize_legend_ratio = require("./sanitize-legend-ratio.js");
var import_constants = require("../../../constants.js");
var import_constants2 = require("../types/constants.js");
const getChartLayoutSettings = (containerSizes, children, isEmpty) => {
  const { provider, graphArea, legendArea, legendProps } = (0, import_extract_layout_slots.extractLayoutSlots)(children);
  const containerAvailableArea = {
    width: containerSizes.width,
    height: containerSizes.height
  };
  if (!legendArea) {
    return {
      graphArea,
      containerAvailableArea,
      legendSettings: {
        orientation: import_constants.ORIENTATION.Vertical,
        showResizer: false,
        hidden: true,
        ratio: 0,
        ratioBoundaries: import_constants2.DEFAULT_RATIO_BOUNDARIES
      },
      provider
    };
  }
  const {
    position,
    // eslint-disable-next-line @typescript-eslint/no-deprecated
    ratio = import_constants.defaultRatio,
    initialRatio,
    onResize,
    overrideRatio,
    ratioBoundaries = import_constants2.DEFAULT_RATIO_BOUNDARIES
  } = legendProps ?? import_constants2.DEFAULT_LEGEND_CONFIG;
  const orientation = (0, import_orientation.getOrientation)(
    containerAvailableArea.width,
    containerAvailableArea.height,
    position
  );
  return {
    graphArea,
    legendArea,
    containerAvailableArea,
    legendSettings: {
      orientation,
      showResizer: !isEmpty,
      hidden: (0, import_lodash_es.isUndefined)(legendArea),
      ratio: (0, import_sanitize_legend_ratio.sanitizeLegendRatio)(ratio, ratioBoundaries, orientation),
      initialRatio: initialRatio ? (0, import_sanitize_legend_ratio.sanitizeLegendRatio)(initialRatio, ratioBoundaries, orientation) : void 0,
      overrideRatio,
      ratioBoundaries,
      onResize
    },
    provider
  };
};
