// @dynatrace/strato-components/charts/timeseries/context/TimeseriesData.context.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TimeseriesData_context_exports = {};
__export(TimeseriesData_context_exports, {
  TimeseriesDataContext: () => TimeseriesDataContext
});
module.exports = __toCommonJS(TimeseriesData_context_exports);
var import_react = require("react");
const TimeseriesDataContext = (0, import_react.createContext)({
  dataPerVariant: {
    area: [],
    bar: [],
    band: [],
    line: [],
    shapePriority: []
  },
  data: [],
  units: [],
  valueRepPerVariant: /* @__PURE__ */ new Map(),
  seriesMap: /* @__PURE__ */ new WeakMap()
});
TimeseriesDataContext["displayName"] = "TimeseriesDataContext";
