// @dynatrace/strato-components/esm/charts/categorical-bar/components/thresholds/CategoricalThresholdBar.js
import { jsx } from "react/jsx-runtime";
import { ThresholdBar } from "../../../core/components/threshold-bar/ThresholdBar.js";
import { getMarkerPixel } from "../../../core/components/threshold-bar/utils/get-marker-pixel.js";
import { getMarkerValue } from "../../../core/components/threshold-bar/utils/get-marker-value.js";
import { isBoundedRange } from "../../../core/types/type-guards.js";
import { useSetState } from "../../hooks/store/selectors.js";
import { useCategoricalBarChartBaseConfig } from "../../hooks/useCategoricalBarChartBaseConfig.js";
import { useCategoricalBarChartScales } from "../../hooks/useCategoricalBarChartScales.js";
import { useCategoricalBarChartThresholds } from "../../hooks/useCategoricalBarChartThresholds.js";
import { useFormatter } from "../../hooks/useFormatter.js";
const CategoricalThresholdBar = ({
  length
}) => {
  const { layout } = useCategoricalBarChartBaseConfig();
  const { valueScale } = useCategoricalBarChartScales();
  const setState = useSetState();
  const formatter = useFormatter();
  const thresholds = useCategoricalBarChartThresholds();
  const handleHover = (id) => {
    setState((prev) => ({
      ...prev,
      hoveredThresholdId: id
    }));
  };
  const getMarkerProps = (yFirst, yLast) => ({
    y0: getMarkerPixel(yFirst, valueScale, layout),
    y1: getMarkerPixel(yLast, valueScale, layout),
    value: getMarkerValue({ y0: yFirst, y1: yLast }, formatter)
  });
  const markers = thresholds.map((threshold) => {
    const domainMin = isBoundedRange(threshold.data) ? threshold.data.value : threshold.data.min ?? -Infinity;
    const domainMax = isBoundedRange(threshold.data) ? threshold.data.value : threshold.data.max ?? Infinity;
    return {
      ...threshold,
      ...getMarkerProps(domainMin, domainMax)
    };
  });
  return markers.length ? /* @__PURE__ */ jsx(
    ThresholdBar,
    {
      markers,
      length,
      layout,
      onHover: handleHover
    }
  ) : null;
};
CategoricalThresholdBar["displayName"] = "TimeseriesThresholdBar";
export {
  CategoricalThresholdBar
};
