// @dynatrace/strato-components/charts/xy-chart/components/tooltip/ClosestPointTooltip/closestPointTooltipBodyUtils.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var closestPointTooltipBodyUtils_exports = {};
__export(closestPointTooltipBodyUtils_exports, {
  buildClosestPointTooltipBodyProps: () => buildClosestPointTooltipBodyProps
});
module.exports = __toCommonJS(closestPointTooltipBodyUtils_exports);
var import_safe_get = require("../../../../core/utils/safe-get.js");
var import_type_guards = require("../../../utils/type-guards.js");
var import_value_representation = require("../../../utils/value-representation.js");
var import_get_tooltip_formatter = require("../utils/get-tooltip-formatter.js");
var import_getClosestPointTooltipBodyContent = require("./hooks/getClosestPointTooltipBodyContent.js");
const buildClosestPointTooltipBodyProps = ({
  datapoint,
  series,
  seriesToMetadata,
  axisIdToFormatter,
  axisIdToScaleDomain,
  datapointMetadataMap,
  intl
}) => {
  const seriesMetadata = (0, import_safe_get.safeGet)(seriesToMetadata, series);
  const enrichedDatapoint = {
    series,
    datapoint,
    metadata: seriesMetadata
  };
  const { xAxis, yAxis } = enrichedDatapoint.metadata;
  const xFormatter = (0, import_get_tooltip_formatter.getTooltipFormatter)(xAxis, axisIdToFormatter, intl);
  const yFormatter = (0, import_get_tooltip_formatter.getTooltipFormatter)(yAxis, axisIdToFormatter, intl);
  const datapointMetadata = (0, import_safe_get.safeGet)(datapointMetadataMap, datapoint);
  const binDomain = datapointMetadata?.binDomain;
  let xDomain = (0, import_safe_get.safeGet)(axisIdToScaleDomain, xAxis.id);
  let yDomain = (0, import_safe_get.safeGet)(axisIdToScaleDomain, yAxis.id);
  if ((0, import_type_guards.isAreaTooltipContext)(enrichedDatapoint)) {
    const { isStacked } = enrichedDatapoint.metadata;
    if (isStacked && datapointMetadata.baseline !== "x" && binDomain) {
      xDomain = { min: binDomain * -1, max: binDomain };
    }
    if (isStacked && datapointMetadata.baseline !== "y" && binDomain) {
      yDomain = { min: binDomain * -1, max: binDomain };
    }
  }
  const { text: textX, duration: durationX } = (0, import_getClosestPointTooltipBodyContent.getClosestPointTooltipBodyContent)(
    [datapoint.x0, datapoint.x1],
    xFormatter,
    xAxis.isRelative ? (value) => (0, import_value_representation.toRelative)(value, xDomain) : void 0
  );
  const { text: textY, duration: durationY } = (0, import_getClosestPointTooltipBodyContent.getClosestPointTooltipBodyContent)(
    [datapoint.y0, datapoint.y1],
    yFormatter,
    yAxis.isRelative ? (value) => (0, import_value_representation.toRelative)(value, yDomain) : void 0
  );
  return {
    xAxisLabel: xAxis.label,
    yAxisLabel: yAxis.label,
    xAxisValue: textX,
    yAxisValue: textY,
    xAxisDuration: durationX,
    yAxisDuration: durationY
  };
};
