// @dynatrace/strato-components/charts/xy-chart/components/tooltip/ClosestPointTooltip/ClosestPointTooltip.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ClosestPointTooltip_exports = {};
__export(ClosestPointTooltip_exports, {
  ClosestPointTooltip: () => ClosestPointTooltip
});
module.exports = __toCommonJS(ClosestPointTooltip_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_intl = require("react-intl");
var import_closestPointTooltipBodyUtils = require("./closestPointTooltipBodyUtils.js");
var import_closestPointTooltipHeaderUtils = require("./closestPointTooltipHeaderUtils.js");
var import_OverlayTooltip = require("../../../../core/components/overlay-tooltip/OverlayTooltip.js");
var import_useAreaData = require("../../../hooks/useAreaData.js");
var import_useAxesFormatter = require("../../../hooks/useAxesFormatter.js");
var import_useXYChartData = require("../../../hooks/useXYChartData.js");
var import_useXYChartTruncationMode = require("../../../hooks/useXYChartTruncationMode.js");
function ClosestPointTooltip({
  datapoint,
  series,
  seriesToMetadata
}) {
  const axisIdToFormatter = (0, import_useAxesFormatter.useAxesFormatter)();
  const { axisIdToScaleDomain, datapointMetadataMap } = (0, import_useXYChartData.useXYChartData)();
  const intl = (0, import_react_intl.useIntl)();
  const { filteredSeriesWithDatapoints } = (0, import_useAreaData.useAreaData)();
  const { truncationMode } = (0, import_useXYChartTruncationMode.useXYChartTruncationMode)();
  const { name, color, actions, shape } = (0, import_closestPointTooltipHeaderUtils.buildClosestPointTooltipHeaderProps)({
    datapoint,
    series,
    seriesToMetadata,
    seriesWithDatapoints: filteredSeriesWithDatapoints,
    datapointMetadataMap
  });
  const { xAxisValue, yAxisValue, xAxisDuration, yAxisDuration } = (0, import_closestPointTooltipBodyUtils.buildClosestPointTooltipBodyProps)({
    datapoint,
    series,
    seriesToMetadata,
    axisIdToFormatter,
    axisIdToScaleDomain,
    datapointMetadataMap,
    intl
  });
  const data = [
    {
      name,
      value: `${yAxisValue} ${yAxisDuration ? `(${yAxisDuration})` : ""}`,
      color,
      shape
    }
  ];
  const label = datapoint.x0 instanceof Date ? "" : void 0;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_OverlayTooltip.OverlayTooltip, { truncation: truncationMode, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_OverlayTooltip.OverlayTooltip.Header,
      {
        header1: xAxisValue,
        duration1: xAxisDuration,
        label
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_OverlayTooltip.OverlayTooltip.Series,
      {
        data,
        variant: "shared",
        activeItem: -1,
        seriesItemTemplate: (item) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayTooltip.OverlayTooltip.SeriesItem, { ...item, actions })
      }
    )
  ] });
}
