// @dynatrace/strato-components/charts/xy-chart/components/tooltip/ClosestPointTooltip/closestPointTooltipHeaderUtils.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var closestPointTooltipHeaderUtils_exports = {};
__export(closestPointTooltipHeaderUtils_exports, {
  buildClosestPointTooltipHeaderProps: () => buildClosestPointTooltipHeaderProps
});
module.exports = __toCommonJS(closestPointTooltipHeaderUtils_exports);
var import_useDefaultAction = require("../../../../core/components/series-action/hooks/useDefaultAction.js");
var import_safe_get = require("../../../../core/utils/safe-get.js");
var import_build_series_action_item = require("../../../utils/build-series-action-item.js");
var import_type_guards = require("../../../utils/type-guards.js");
const buildClosestPointTooltipHeaderProps = ({
  datapoint,
  series,
  seriesToMetadata,
  seriesWithDatapoints,
  datapointMetadataMap,
  truncationMode,
  idFactory = () => crypto.randomUUID()
}) => {
  const metadata = (0, import_safe_get.safeGet)(seriesToMetadata, series);
  const enrichedDatapoint = {
    series,
    datapoint,
    metadata
  };
  const { name = "", color = { Default: "blue" } } = metadata;
  const shape = "shape" in metadata ? metadata.shape : void 0;
  const content = {
    id: idFactory(),
    name
  };
  let actions = void 0;
  if ((0, import_type_guards.isAreaTooltipContext)(enrichedDatapoint)) {
    const { actions: seriesActions } = enrichedDatapoint.metadata;
    actions = seriesActions ? () => {
      const actionItem = (0, import_build_series_action_item.buildAreaSeriesActionItem)({
        datapoint: enrichedDatapoint.datapoint,
        allSeries: seriesWithDatapoints,
        seriesToMetadata,
        datapointMetadataMap,
        currentSeries: enrichedDatapoint.series
      });
      return seriesActions(actionItem);
    } : void 0;
  } else {
    const { actions: seriesActions } = enrichedDatapoint.metadata;
    actions = () => seriesActions?.({
      ...datapoint,
      name
    });
  }
  return {
    color: color.Default,
    name,
    actions,
    defaultActions: (0, import_useDefaultAction.getDefaultAction)(content),
    truncationMode,
    shape
  };
};
