// @dynatrace/strato-components/charts/pie/utils/build-group-threshold.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var build_group_threshold_exports = {};
__export(build_group_threshold_exports, {
  buildGroupThreshold: () => buildGroupThreshold
});
module.exports = __toCommonJS(build_group_threshold_exports);
var import_lodash_es = require("lodash");
var import_build_slices = require("./build-slices.js");
var import_default = require("../types/default.js");
function buildGroupThreshold(data, grouping, colorPalette, unit) {
  const internalSlicesData = (0, import_build_slices.buildSlices)(data, colorPalette, unit);
  const threshold = grouping?.threshold;
  if (threshold?.type === import_default.THRESHOLD_TYPE.NUMBER_OF_SLICES) {
    return calculateNumberOfSlicesGrouping(
      internalSlicesData,
      threshold,
      grouping?.name
    );
  }
  if (threshold?.type === import_default.THRESHOLD_TYPE.RELATIVE) {
    return calculateRelativeGrouping(
      internalSlicesData,
      threshold,
      grouping?.name
    );
  }
  if (threshold?.type === import_default.THRESHOLD_TYPE.ABSOLUTE) {
    return calculateAbsoluteGrouping(
      internalSlicesData,
      threshold,
      grouping?.name
    );
  }
  return internalSlicesData;
}
function calculateNumberOfSlicesGrouping(data, threshold, groupingName) {
  const slicesWithValue = data.filter((slice) => slice.value);
  const slicesToMerge = slicesWithValue.slice(threshold.value - 1);
  const remainingSlices = slicesWithValue.slice(0, threshold.value - 1);
  if (threshold.value === 0) {
    return [...slicesToMerge, ...remainingSlices];
  }
  return generateGroup(slicesToMerge, remainingSlices, groupingName);
}
function calculateRelativeGrouping(data, threshold, groupingName) {
  const slicesWithValue = data.filter((slice) => slice.value);
  const [slicesToMerge, remainingSlices] = (0, import_lodash_es.partition)(
    slicesWithValue,
    (slice) => slice.relativeValue < threshold.value
  );
  return generateGroup(slicesToMerge, remainingSlices, groupingName);
}
function calculateAbsoluteGrouping(data, threshold, groupingName) {
  const slicesWithValue = data.filter((slice) => slice.value);
  const [slicesToMerge, remainingSlices] = (0, import_lodash_es.partition)(
    slicesWithValue,
    (slice) => slice.value < threshold.value
  );
  return generateGroup(slicesToMerge, remainingSlices, groupingName);
}
function generateGroup(slicesToMerge, data, groupingName) {
  const isWithoutGrouping = !slicesToMerge.length || slicesToMerge.length === 1;
  if (isWithoutGrouping) {
    return [...data, ...slicesToMerge];
  }
  const groupIndexValue = slicesToMerge[0].id.split("-")[1];
  const groupId = groupingName ? groupingName + "-" + groupIndexValue : "other-" + groupIndexValue;
  const mergedSlice = slicesToMerge.reduce(
    (acc, curr) => ({
      ...acc,
      ...curr,
      category: groupingName ?? "Other",
      value: acc.value + curr.value,
      relativeValue: acc.relativeValue + curr.relativeValue,
      color: slicesToMerge[0].color,
      id: groupId
    }),
    {
      value: 0,
      relativeValue: 0
    }
  );
  return [...data, mergedSlice];
}
