// @dynatrace/strato-components/charts/pie/utils/get-tooltip-content.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var get_tooltip_content_exports = {};
__export(get_tooltip_content_exports, {
  getTooltipContent: () => getTooltipContent
});
module.exports = __toCommonJS(get_tooltip_content_exports);
const getTooltipContent = (sliceId, data, handlerCatch, formatter) => {
  const formattedValue = handlerCatch(() => formatter(data.value, data.unit));
  const relativeValue = data.relativeValue.toFixed(2);
  return {
    id: sliceId,
    name: data.category,
    color: data.color.Default,
    absoluteValue: formattedValue,
    relativeValue: Number(relativeValue)
  };
};
