// @dynatrace/strato-components/charts/pie/utils/group-slices-thinner-than-pad-angle.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var group_slices_thinner_than_pad_angle_exports = {};
__export(group_slices_thinner_than_pad_angle_exports, {
  groupSlicesThinnerThanPadAngle: () => groupSlicesThinnerThanPadAngle
});
module.exports = __toCommonJS(group_slices_thinner_than_pad_angle_exports);
var import_plot_data_utils = require("./plot-data.utils.js");
const groupSlicesThinnerThanPadAngle = (slices, padAngle) => {
  let slicesThinnerThanPadAngleValueSum = 0;
  let totalAbsoluteValueSum = 0;
  const newSlices = [];
  const padAngleDeg = padAngle * 180 / Math.PI;
  for (const slice of slices) {
    const sliceAngleDeg = slice.relativeValue / 100 * 360;
    if (sliceAngleDeg <= padAngleDeg) {
      slicesThinnerThanPadAngleValueSum += slice.value;
    } else {
      newSlices.push(slice);
    }
    totalAbsoluteValueSum += slice.value;
  }
  if (slicesThinnerThanPadAngleValueSum > 0) {
    const negativeSlicesAggregatedSliceColorToken = {
      Default: "transparent"
    };
    const negativeSlicesAggregatedSlice = {
      category: "",
      value: slicesThinnerThanPadAngleValueSum,
      color: negativeSlicesAggregatedSliceColorToken,
      id: "thinner-slices",
      relativeValue: (0, import_plot_data_utils.getSlicePercentage)(
        slicesThinnerThanPadAngleValueSum,
        totalAbsoluteValueSum
      )
    };
    newSlices.push(negativeSlicesAggregatedSlice);
  }
  return newSlices;
};
