// @dynatrace/strato-components/charts/pie/utils/iterate-slots.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var iterate_slots_exports = {};
__export(iterate_slots_exports, {
  iterateSlots: () => iterateSlots
});
module.exports = __toCommonJS(iterate_slots_exports);
var import_react = require("react");
var import_Intent = require("../../../core/slots/Intent/Intent.js");
var import_is_slot = require("../../../core/utils/is-slot.js");
var import_ColorRule = require("../../core/slots/ColorRule.js");
var import_ChartToolbar = require("../../core/slots/toolbar/ChartToolbar.js");
var import_CompactToolbar = require("../../core/slots/toolbar/CompactToolbar.js");
var import_DownloadData = require("../../core/slots/toolbar/DownloadData.js");
var import_extract_behavior_tracking_props = require("../../core/utils/extract-behavior-tracking-props.js");
var import_Toolbar = require("../shared/slots/Toolbar.js");
var import_DonutChartDownloadCSV = require("../slots/DonutChartDownloadCSV.js");
var import_DonutChartGrouping = require("../slots/DonutChartGrouping.js");
var import_DonutChartInner = require("../slots/DonutChartInner.js");
var import_DonutChartLabels = require("../slots/DonutChartLabels.js");
var import_DonutChartLegend = require("../slots/DonutChartLegend.js");
var import_DonutChartTooltip = require("../slots/DonutChartTooltip.js");
var import_PieChartDownloadCSV = require("../slots/PieChartDownloadCSV.js");
var import_PieChartGrouping = require("../slots/PieChartGrouping.js");
var import_PieChartLabels = require("../slots/PieChartLabels.js");
var import_PieChartLegend = require("../slots/PieChartLegend.js");
var import_PieChartTooltip = require("../slots/PieChartTooltip.js");
const iterateChartToolbarSlots = (children) => {
  let slots = {};
  import_react.Children.forEach(children, (childInteractions) => {
    if ((0, import_is_slot.isSlot)(childInteractions, import_DownloadData.DownloadData)) {
      slots = { ...slots, downloadData: { enabled: true } };
    }
  });
  return slots;
};
function iterateSlots(children) {
  let slots = {
    labels: {},
    grouping: {},
    legend: {},
    toolbar: void 0,
    innerElement: void 0,
    colorRules: [],
    tooltip: void 0
  };
  let intents;
  import_react.Children.map(children, (child) => {
    if ((0, import_is_slot.isSlot)(child, import_DonutChartLabels.Labels) || (0, import_is_slot.isSlot)(child, import_PieChartLabels.Labels)) {
      slots = { ...slots, labels: child.props };
    } else if ((0, import_is_slot.isSlot)(child, import_PieChartGrouping.Grouping) || (0, import_is_slot.isSlot)(child, import_DonutChartGrouping.Grouping)) {
      slots = { ...slots, grouping: child.props };
    } else if ((0, import_is_slot.isSlot)(child, import_PieChartLegend.Legend) || (0, import_is_slot.isSlot)(child, import_DonutChartLegend.Legend)) {
      slots = { ...slots, legend: child.props };
    } else if ((0, import_is_slot.isSlot)(child, import_Toolbar.Toolbar)) {
      slots = {
        ...slots,
        toolbar: {
          ...slots.toolbar,
          ...(0, import_CompactToolbar.sanitizeProps)(child.props),
          ...(0, import_extract_behavior_tracking_props.extractBehaviorTrackingProps)(child.props)
        }
      };
    } else if ((0, import_is_slot.isSlot)(child, import_ChartToolbar.ChartToolbar)) {
      const { children: children2, ...toolbarProps } = child.props;
      const toolbarSlots = iterateChartToolbarSlots(children2);
      const toolbar = { ...toolbarProps, ...toolbarSlots };
      slots = { ...slots, toolbar };
    } else if ((0, import_is_slot.isSlot)(child, import_DonutChartDownloadCSV.DonutChartDownloadCSVSlot) || (0, import_is_slot.isSlot)(child, import_PieChartDownloadCSV.PieChartDownloadCSVSlot)) {
      slots = {
        ...slots,
        toolbar: {
          ...slots.toolbar,
          downloadData: { enabled: true }
        }
      };
    } else if ((0, import_is_slot.isSlot)(child, import_Intent.Intent)) {
      intents = intents ? [...intents, child.props] : [child.props];
    } else if ((0, import_is_slot.isSlot)(child, import_DonutChartInner.DonutChartInner)) {
      const inner = child.props.children;
      slots = { ...slots, innerElement: inner };
    } else if ((0, import_is_slot.isSlot)(child, import_ColorRule.ColorRule)) {
      (0, import_ColorRule.isValidColorRule)(child.props) && slots.colorRules.push(child.props);
    } else if ((0, import_is_slot.isSlot)(child, import_PieChartTooltip.Tooltip) || (0, import_is_slot.isSlot)(child, import_DonutChartTooltip.Tooltip)) {
      slots = {
        ...slots,
        tooltip: {
          hidden: child.props.hidden ?? false,
          template: child.props.children
        }
      };
    }
  });
  if (intents?.length) {
    slots = {
      ...slots,
      toolbar: {
        ...slots.toolbar,
        intents
      }
    };
  }
  return slots;
}
