// @dynatrace/strato-components/charts/pie/utils/pie-config.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var pie_config_exports = {};
__export(pie_config_exports, {
  buildPieChartConfig: () => buildPieChartConfig,
  deserializeConfig: () => deserializeConfig,
  getColorPaletteConfiguration: () => getColorPaletteConfiguration,
  getGroupingConfiguration: () => getGroupingConfiguration,
  getLegendConfiguration: () => getLegendConfiguration
});
module.exports = __toCommonJS(pie_config_exports);
var import_lodash_es = require("lodash");
var import_constants = require("../../core/constants.js");
var import_ColorRule = require("../../core/slots/ColorRule.js");
var import_color_palette = require("../../core/types/color-palette.js");
var import_defaultUnitFormatter = require("../../core/utils/defaultUnitFormatter.js");
var import_is_valid_color_palette = require("../../core/utils/is-valid-color-palette.js");
var import_is_valid_legend_position = require("../../core/utils/is-valid-legend-position.js");
var import_constants2 = require("../constants.js");
var import_default = require("../types/default.js");
var import_is_valid_threshold_type = require("../utils/is-valid-threshold-type.js");
function getColorPaletteConfiguration(contextColorPalette, propsColorPalette) {
  const colorPalette = propsColorPalette ?? contextColorPalette ?? import_default.defaultConfig.colorPalette;
  return (0, import_color_palette.isCustomColorPalette)(colorPalette) || (0, import_is_valid_color_palette.isValidColorPalette)(colorPalette) ? colorPalette : import_default.defaultConfig.colorPalette;
}
function getGroupingConfiguration(contextGrouping, slotsGrouping) {
  let grouping = (0, import_lodash_es.merge)(
    {},
    import_default.defaultConfig.grouping,
    contextGrouping,
    slotsGrouping
  );
  grouping = {
    ...grouping,
    threshold: {
      ...grouping.threshold,
      type: (0, import_is_valid_threshold_type.isValidThresholdType)(grouping.threshold.type) ? grouping.threshold.type : import_constants2.THRESHOLD_TYPE_OPTION.Relative
    }
  };
  return grouping;
}
function getLegendConfiguration(contextLegendConfig, slotsLegendConfig) {
  let legend = (0, import_lodash_es.merge)(
    {},
    import_default.defaultConfig.legend,
    contextLegendConfig,
    slotsLegendConfig
  );
  const position = legend.position && (0, import_is_valid_legend_position.isValidLegendPosition)(legend.position) ? legend.position : import_default.defaultConfig.legend.position;
  legend = {
    ...legend,
    position
  };
  return legend;
}
const validateToolbarContextConfig = (contextConfig) => !(0, import_lodash_es.isUndefined)(contextConfig) && contextConfig.downloadData?.enabled;
const getToolbarConfiguration = (contextConfig, slots) => {
  const hasValidToolbarConfig = validateToolbarContextConfig(contextConfig) || !(0, import_lodash_es.isUndefined)(slots);
  return hasValidToolbarConfig ? (0, import_lodash_es.merge)({}, contextConfig, slots) : void 0;
};
function buildPieChartConfig(props, contextConfig, slots) {
  const colorPalette = getColorPaletteConfiguration(
    contextConfig.colorPalette,
    props.colorPalette
  );
  const grouping = getGroupingConfiguration(
    contextConfig.grouping,
    slots.grouping
  );
  const legend = getLegendConfiguration(contextConfig.legend, slots.legend);
  const formatter = props.formatter ?? contextConfig.formatter ?? (0, import_defaultUnitFormatter.defaultUnitFormatter)(props.data.unit, 0);
  const truncationMode = props.truncationMode ?? contextConfig.truncationMode ?? import_constants.DEFAULT_TRUNCATION_MODE;
  const toolbar = getToolbarConfiguration(contextConfig.toolbar, slots.toolbar);
  const seriesActions = props.seriesActions ?? contextConfig.seriesActions ?? void 0;
  const colorRules = (contextConfig.colorRules || []).filter(import_ColorRule.isValidColorRule).concat(slots.colorRules || []);
  return {
    height: props.height ?? import_default.defaultConfig.height,
    colorPalette,
    labels: (0, import_lodash_es.merge)({}, import_default.defaultConfig.labels, contextConfig.labels, slots.labels),
    grouping,
    legend,
    formatter,
    truncationMode,
    toolbar,
    innerElement: slots.innerElement,
    seriesActions,
    colorRules,
    tooltip: slots.tooltip ?? contextConfig.tooltip
  };
}
function deserializeConfig(serializedConfig) {
  try {
    return (0, import_lodash_es.merge)({}, import_default.defaultConfig, JSON.parse(serializedConfig));
  } catch {
    return import_default.defaultConfig;
  }
}
