// @dynatrace/strato-components/content/microguide/Microguide.sty.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Microguide_css_exports = {};
__export(Microguide_css_exports, {
  animatedContent: () => animatedContent,
  contentWrapper: () => contentWrapper,
  footer: () => footer,
  stepContainer: () => stepContainer,
  visualWrapper: () => visualWrapper,
  wrapper: () => wrapper,
  wrapperPosition: () => wrapperPosition
});
module.exports = __toCommonJS(Microguide_css_exports);
var import_Microguide_css_ts_vanilla = require("./Microguide.css");
var animatedContent = "_z4mh733-3-6-0";
var contentWrapper = "_z4mh732-3-6-0";
var footer = "_z4mh734-3-6-0";
var stepContainer = "_z4mh730-3-6-0";
var visualWrapper = "_z4mh731-3-6-0";
var wrapper = "_z4mh735-3-6-0";
var wrapperPosition = { "bottom-left": "_z4mh736-3-6-0", "top-right": "_z4mh737-3-6-0", "top-left": "_z4mh738-3-6-0", "bottom-right": "_z4mh739-3-6-0" };
