// @dynatrace/strato-components/content/microguide/Microguide.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Microguide_exports = {};
__export(Microguide_exports, {
  Microguide: () => Microguide
});
module.exports = __toCommonJS(Microguide_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_dom = require("react-dom");
var import_react_intl = require("react-intl");
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
var import_MicroguideContext = require("./contexts/MicroguideContext.js");
var styles = __toESM(require("./Microguide.sty.js"));
var import_Step = require("./Step.js");
var import_Button = require("../../buttons/button/Button.js");
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useOverlayStacking = require("../../core/hooks/useOverlayStacking.js");
var import_useSafeForwardProps = require("../../core/hooks/useSafeForwardProps.js");
var import_useSlots = require("../../core/hooks/useSlots.js");
var import_Flex = require("../../layouts/flex/Flex.js");
var import_Surface = require("../../layouts/surface/Surface.js");
var import_useKeyboardFocusHandling = require("../../overlays/hooks/useKeyboardFocusHandling.js");
var import_Content = require("../shared-components/Content.js");
var import_Title = require("../shared-components/Title.js");
var import_Visual = require("../shared-components/Visual.js");
const COMPONENT_NAME = "Microguide";
const MicroguideRoot = (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const {
      open,
      placement = "bottom-right",
      onClose,
      children,
      onStepChange,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      className: consumerClassName,
      style: consumerStyle,
      ...remainingProps
    } = props;
    const microguideContainer = (0, import_react.useRef)(null);
    const closeButtonRef = (0, import_react.useRef)(null);
    const [currentStep, setCurrentStep] = (0, import_react.useState)(0);
    const [isNext, setIsNext] = (0, import_react.useState)(true);
    const [hasTitle, setHasTitle] = (0, import_react.useState)(false);
    const [hasContent, setHasContent] = (0, import_react.useState)(false);
    const [handleFocusOnKeydown, closedOnKeydown] = (0, import_useKeyboardFocusHandling.useKeyboardFocusHandling)(closeButtonRef);
    const overlayStacking = (0, import_useOverlayStacking.useOverlayStacking)(void 0, "microguide");
    const intl = (0, import_react_intl.useIntl)();
    const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
    const ariaLabelingProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(restProps);
    const forwardProps = (0, import_useSafeForwardProps.useSafeForwardProps)(restProps, ariaLabelingProps);
    const [slots] = (0, import_useSlots.useSlots)(children, {
      steps: { match: import_Step.Step, multiple: true }
    });
    const stepsLength = slots.steps?.length ?? 0;
    const isLastStep = currentStep + 1 >= stepsLength;
    const nextGotItDataDtProperties = `state:${isLastStep ? "gotIt" : "next"}`;
    if (hasTitle && ariaLabelingProps["aria-labelledby"] === void 0) {
      ariaLabelingProps["aria-labelledby"] = "strato-microguide-title";
    }
    if (hasContent && ariaLabelingProps["aria-describedby"] === void 0) {
      ariaLabelingProps["aria-describedby"] = "strato-microguide-content";
    }
    const closeOnEscape = (event) => {
      if (event.key !== "Escape") {
        return;
      }
      event.preventDefault();
      closedOnKeydown.current = true;
      onClose(currentStep, isLastStep);
    };
    (0, import_react.useLayoutEffect)(() => {
      const container = document.createElement("div");
      container.setAttribute("aria-live", "polite");
      container.setAttribute("data-microguide-container", "true");
      microguideContainer.current = container;
      const toastContainer = document.querySelector("[data-toast-container]");
      if (toastContainer?.parentNode) {
        toastContainer.parentNode.insertBefore(container, toastContainer);
      } else {
        document.body.appendChild(container);
      }
      return () => {
        microguideContainer.current?.remove();
      };
    }, []);
    (0, import_react.useEffect)(() => {
      if (open === false) {
        setCurrentStep(0);
        handleFocusOnKeydown();
      }
    }, [handleFocusOnKeydown, open]);
    const handleBack = () => {
      const previousStep = currentStep - 1;
      setIsNext(false);
      onStepChange?.(previousStep);
      setCurrentStep(previousStep);
    };
    const handleNext = () => {
      const nextStep = currentStep + 1;
      setIsNext(true);
      onStepChange?.(nextStep);
      setCurrentStep(nextStep);
    };
    return open ? (0, import_react_dom.createPortal)(
      /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
        import_Surface.Surface,
        {
          ref: forwardedRef,
          elevation: "floating",
          onKeyDown: closeOnEscape,
          role: "dialog",
          "aria-modal": "false",
          className: (0, import_clsx.default)(
            styles.wrapper,
            styles.wrapperPosition[placement],
            consumerClassName
          ),
          style: {
            // since the surface applies `position: relative` by default,
            // we ensure higher specificity by applying position as an inline style
            position: "fixed",
            zIndex: overlayStacking.style.zIndex,
            ...consumerStyle
          },
          "data-testid": dataTestId,
          "data-dtrum-mask": dataDtrumMask,
          "data-dtrum-allow": dataDtrumAllow,
          ...ariaLabelingProps,
          ...behavioralTrackingProps,
          ...forwardProps,
          "data-dt-component": COMPONENT_NAME,
          children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_MicroguideContext.MicroguideContext.Provider,
              {
                value: {
                  onClose,
                  currentStep,
                  totalSteps: stepsLength,
                  isNext,
                  setHasContent,
                  setHasTitle,
                  closeButtonRef,
                  closedOnKeydown
                },
                children: slots.steps?.[currentStep]
              }
            ),
            /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
              import_Flex.Flex,
              {
                className: styles.footer,
                flexDirection: "row",
                justifyContent: "space-between",
                alignItems: "center",
                gap: 16,
                paddingX: 16,
                paddingBottom: 16,
                paddingTop: 8,
                children: [
                  /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { style: { color: import_colors.default.Text.Neutral.Subdued }, children: stepsLength > 1 && `${currentStep + 1}/${stepsLength}` }),
                  /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { gap: 8, children: [
                    currentStep > 0 && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button, { onClick: handleBack, "data-dt-name": "Back", children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                      import_react_intl.FormattedMessage,
                      {
                        description: "Microguide Back button.",
                        defaultMessage: "Back",
                        id: "N27hQ6LyGqdym2Yi"
                      }
                    ) }),
                    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                      import_Button.Button,
                      {
                        variant: isLastStep ? "accent" : "emphasized",
                        color: isLastStep ? "primary" : "neutral",
                        onClick: isLastStep ? () => onClose(currentStep, true) : handleNext,
                        onKeyDown: () => closedOnKeydown.current = isLastStep,
                        "data-dt-name": "NextGotIt",
                        "data-dt-properties": nextGotItDataDtProperties,
                        children: isLastStep ? intl.formatMessage({
                          defaultMessage: "Got it",
                          id: "fzkORq3AsvX/NtMB",
                          description: "Microguide Got it button at last step."
                        }) : intl.formatMessage({
                          defaultMessage: "Next",
                          id: "+jn6HNnzB4IDIbZu",
                          description: "Microguide Next button."
                        })
                      }
                    )
                  ] })
                ]
              }
            )
          ]
        }
      ),
      microguideContainer.current ?? document.body
    ) : null;
  }
);
MicroguideRoot.displayName = COMPONENT_NAME;
const Microguide = Object.assign(MicroguideRoot, {
  Step: import_Step.Step,
  Title: import_Title.Title,
  Visual: import_Visual.Visual,
  Content: import_Content.Content
});
