// @dynatrace/strato-components/content/microguide/Step.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Step_exports = {};
__export(Step_exports, {
  Step: () => Step
});
module.exports = __toCommonJS(Step_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_framer_motion = require("framer-motion");
var import_react2 = require("react");
var import_react_intl = require("react-intl");
var import_animations_js = __toESM(require("@dynatrace/strato-design-tokens/animations-js"));
var import_strato_icons = require("@dynatrace/strato-icons");
var import_useMicroguideContext = require("./hooks/useMicroguideContext.js");
var styles = __toESM(require("./Microguide.sty.js"));
var import_Button = require("../../buttons/button/Button.js");
var import_useSlots = require("../../core/hooks/useSlots.js");
var import_framer_motion_easings = require("../../core/utils/framer-motion-easings.js");
var import_Flex = require("../../layouts/flex/Flex.js");
var import_Content = require("../shared-components/Content.js");
var import_Title = require("../shared-components/Title.js");
var import_Visual = require("../shared-components/Visual.js");
const contentVariants = {
  enter: (isNext) => {
    return {
      x: isNext ? "100%" : "-100%",
      opacity: 0
    };
  },
  center: {
    x: 0,
    opacity: 1,
    position: "relative"
  },
  exit: (isNext) => {
    return {
      x: isNext ? "-100%" : "100%",
      opacity: 0,
      position: "absolute",
      left: 0
    };
  }
};
const fadeProps = {
  initial: {
    opacity: 0
  },
  animate: {
    opacity: 1,
    position: "relative"
  },
  exit: {
    opacity: 0,
    position: "absolute",
    left: 0
  }
};
const Step = ({ children }) => {
  const {
    onClose,
    currentStep,
    totalSteps,
    isNext,
    setHasContent,
    setHasTitle,
    closeButtonRef,
    closedOnKeydown
  } = (0, import_useMicroguideContext.useMicroguideContext)();
  const intl = (0, import_react_intl.useIntl)();
  const shouldReduceMotion = (0, import_framer_motion.useReducedMotion)();
  const transitionProp = {
    duration: shouldReduceMotion ? 0 : import_animations_js.default.SlideIn.Duration.Base,
    ease: import_framer_motion_easings.moveEasing
  };
  const [slots] = (0, import_useSlots.useSlots)(children, {
    title: import_Title.Title,
    visual: import_Visual.Visual,
    content: import_Content.Content
  });
  (0, import_react2.useEffect)(() => {
    setHasTitle(slots.title !== void 0);
    setHasContent(slots.content !== void 0);
  }, [setHasContent, setHasTitle, slots.content, slots.title]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_Flex.Flex,
    {
      as: import_framer_motion.motion.div,
      flexDirection: "column",
      gap: 8,
      className: styles.stepContainer,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
          import_Flex.Flex,
          {
            flexDirection: "row",
            justifyContent: "space-between",
            alignItems: "flex-start",
            gap: 16,
            paddingTop: 8,
            paddingLeft: 16,
            paddingRight: 8,
            children: [
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_Flex.Flex,
                {
                  flexItem: true,
                  paddingY: 6,
                  flexGrow: 1,
                  id: "strato-microguide-title",
                  style: { position: "relative" },
                  children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_framer_motion.AnimatePresence, { initial: false, children: slots.title && /* @__PURE__ */ (0, import_react.createElement)(
                    import_framer_motion.motion.div,
                    {
                      ...fadeProps,
                      transition: transitionProp,
                      key: `strato-microguide-title-${currentStep}`
                    },
                    slots.title
                  ) })
                }
              ),
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_Button.Button,
                {
                  onClick: () => onClose(currentStep, currentStep + 1 === totalSteps),
                  onKeyDown: () => closedOnKeydown.current = true,
                  ref: closeButtonRef,
                  "aria-label": intl.formatMessage({
                    defaultMessage: "Close microguide",
                    id: "0yl3EugW5oCHVbO1",
                    description: "Close the microguide."
                  }),
                  "data-dt-name": "CloseX",
                  children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.XmarkIcon, {}) })
                }
              )
            ]
          }
        ),
        slots.visual && /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: styles.visualWrapper, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_framer_motion.AnimatePresence, { initial: false, children: /* @__PURE__ */ (0, import_react.createElement)(
          import_framer_motion.motion.div,
          {
            ...fadeProps,
            transition: transitionProp,
            key: `strato-microguide-visual-${currentStep}`
          },
          slots.visual
        ) }) }),
        slots.content && /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: styles.contentWrapper, id: "strato-microguide-content", children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_framer_motion.AnimatePresence, { initial: false, custom: isNext, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_Flex.Flex,
          {
            as: import_framer_motion.motion.div,
            variants: contentVariants,
            className: styles.animatedContent,
            custom: isNext,
            initial: "enter",
            animate: "center",
            exit: "exit",
            transition: transitionProp,
            flexDirection: "column",
            children: slots.content
          },
          `strato-microguide-content-${currentStep}`
        ) }) })
      ]
    }
  );
};
Step["displayName"] = "Microguide.Step";
