// @dynatrace/strato-components/filters/filter-field/utils/conversion/string-tree-conversion.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var string_tree_conversion_exports = {};
__export(string_tree_conversion_exports, {
  convertStringToFilterFieldTree: () => convertStringToFilterFieldTree
});
module.exports = __toCommonJS(string_tree_conversion_exports);
var import_tree_conversion = require("./tree-conversion.js");
var import_getNumberParser = require("../../parser/getNumberParser.js");
var import_parser = require("../../parser/parser.js");
var import_normalize_validator_map = require("../../utils/validation-utils/normalize-validator-map.js");
function convertStringToFilterFieldTree(filterString, optionsOrValidatorMap, parserConfig) {
  let validatorMap;
  let customTypes = void 0;
  let _parserConfig;
  if (isConvertStringToFilterFieldTreeOptions(optionsOrValidatorMap)) {
    validatorMap = optionsOrValidatorMap?.validatorMap;
    customTypes = optionsOrValidatorMap?.customTypes;
    _parserConfig = optionsOrValidatorMap?.parserConfig;
  } else {
    validatorMap = optionsOrValidatorMap;
    _parserConfig = parserConfig;
  }
  const numberParser = (0, import_getNumberParser.getNumberParser)();
  const parsedTree = import_parser.parser.parse(filterString);
  const { tree, isValid } = (0, import_tree_conversion.convertLezerTree)({
    tree: parsedTree,
    numberParser,
    doc: filterString,
    internalValidatorMap: (0, import_normalize_validator_map.toInternalValidatorMap)(validatorMap),
    customTypes,
    parserConfig: _parserConfig
  });
  return { tree, isValid };
}
function isConvertStringToFilterFieldTreeOptions(value) {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    return false;
  }
  return "validatorMap" in value || "customTypes" in value || "parserConfig" in value;
}
