// @dynatrace/strato-components/filters/filter-field/utils/conversion/tree-string-conversion.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var tree_string_conversion_exports = {};
__export(tree_string_conversion_exports, {
  convertFilterFieldTreeToString: () => convertFilterFieldTreeToString
});
module.exports = __toCommonJS(tree_string_conversion_exports);
var import_getNumberParser = require("../../parser/getNumberParser.js");
function convertFilterFieldTreeToString(tree) {
  return _convertFilterFieldTreeToString(tree, "");
}
function getOperatorInfo(operatorValue, operatorType) {
  const operatorInfo = {
    textValue: operatorValue,
    positionOperator: ""
  };
  switch (operatorValue) {
    case "ends-with":
    case "starts-with":
    case "contains":
      operatorInfo.textValue = "=";
      operatorInfo.positionOperator = operatorValue;
      break;
    case "not-ends-with":
    case "not-starts-with":
    case "not-contains":
      operatorInfo.textValue = "!=";
      operatorInfo.positionOperator = operatorValue.substring(4);
      break;
    case "matches-phrase":
      operatorInfo.textValue = "~";
      break;
    case "not-matches-phrase":
      operatorInfo.textValue = "!~";
      break;
    case "search":
      operatorInfo.textValue = "* ~";
      break;
  }
  if (operatorType === "ExistsOperator") {
    operatorInfo.textValue = operatorValue ? "= *" : "!= *";
  }
  if (operatorType === "SearchOperator") {
    operatorInfo.textValue = "* ~";
  }
  return operatorInfo;
}
function hasToBeWrappedInQuotes(value, isString, hasPositionOperator) {
  const numberParser = (0, import_getNumberParser.getNumberParser)();
  const valueAsNumber = numberParser.parse(value);
  const durationRegex = /^\d+(ns|ms|s|m|h|d)$/;
  return value === "" || (value === "true" || value === "false") && isString || !isNaN(valueAsNumber) && isString && !hasPositionOperator || durationRegex.test(value) && isString && !hasPositionOperator || value.charAt(0) === "$" && isString;
}
function getNestedValue({
  value: _value,
  textValue,
  nodeType,
  isFreeText,
  positionOperator,
  separator
}) {
  let value = _value;
  if (value === void 0) {
    return "";
  }
  if (Array.isArray(value)) {
    let combinedString = "";
    value.forEach((node, index) => {
      return combinedString = _convertFilterFieldTreeToString(
        node,
        separator && index !== 0 ? `${combinedString}${separator}` : combinedString
      );
    });
    return combinedString;
  }
  if (textValue) {
    if (positionOperator === "contains" || positionOperator === "ends-with") {
      textValue = textValue.replaceAll(/(?<=^")\*|^\*/g, "");
    }
    if (positionOperator === "contains" || positionOperator === "starts-with") {
      textValue = textValue.replaceAll(/\*(?="$)|\*$/g, "");
    }
    return textValue;
  }
  value = `${value}`;
  if (nodeType === "Error") {
    if (!isFreeText) {
      return value;
    }
    return value.replaceAll(/\\/g, "\\\\");
  }
  const wrapInQuotes = hasToBeWrappedInQuotes(
    value,
    nodeType === "String",
    positionOperator
  );
  value = value.replaceAll(/\\/g, "\\\\");
  value = value.replaceAll(/"/g, '\\"');
  const positionOperatorInValueRegex = /\*/g;
  value = value.replaceAll(positionOperatorInValueRegex, "\\*");
  const reservedCharactersRegex = /(?!^\$)(?<!\\)(?:\\\\)*[,()!<>="$:[\]~ ]|\\$/;
  if (reservedCharactersRegex.test(value) || wrapInQuotes) {
    return `"${value}"`;
  }
  return value;
}
function _convertFilterFieldTreeToString(tree = void 0, convertedString = "") {
  if (!tree) {
    return "";
  }
  switch (tree.type) {
    case "Group": {
      let groupString = "";
      if (tree.explicit) {
        groupString = "(";
      }
      const childValues = getNestedValue({
        value: tree.children,
        nodeType: tree.type,
        separator: tree.explicitLogicalOperator ? ` ${tree.logicalOperator}` : ""
      });
      groupString = groupString + childValues + `${tree.children.length === 1 && tree.explicitLogicalOperator ? " " + tree.logicalOperator : ""}`;
      if (tree.explicit) {
        const lastChild = tree.children[tree.children.length - 1];
        if (!(lastChild && lastChild.type === "Error" && lastChild.value === "" && tree.children.length !== 1)) {
          groupString += ")";
        }
      }
      convertedString = addNodeToString(convertedString, groupString);
      break;
    }
    case "Error":
      if (tree.value !== "()") {
        convertedString = addNodeToString(
          convertedString,
          getNestedValue({
            value: tree.value,
            textValue: tree.textValue,
            nodeType: tree.type
          })
        );
      }
      break;
    case "Statement": {
      let statement = "";
      const { key, comparisonOperator, value } = tree;
      if (key && key.type === "JSONPath") {
        statement = `${key.value}${comparisonOperator || value ? " " : ""}`;
      } else if (key) {
        statement = `${getNestedValue({
          value: key.value,
          textValue: key.textValue,
          nodeType: key.type,
          isFreeText: !(comparisonOperator ?? value)
        })}${comparisonOperator || value ? " " : ""}`;
      }
      let operatorInfo = { textValue: "", positionOperator: "" };
      if (comparisonOperator) {
        operatorInfo = getOperatorInfo(
          comparisonOperator.value,
          comparisonOperator.type
        );
        statement = `${statement}${operatorInfo.textValue}${value ? " " : ""}`;
      }
      if (value) {
        const val = value.type === "Duration" ? `${value.value}${value.unit}` : value.value;
        const childValues = getNestedValue({
          value: val,
          textValue: value.textValue,
          nodeType: value.type,
          positionOperator: operatorInfo.positionOperator,
          separator: ","
        });
        if (value.type === "List" || value.type === "Error" && Array.isArray(val)) {
          statement = `${statement}(${childValues})`;
        } else if (value.type === "Error") {
          statement = statement + childValues;
        } else {
          statement = statement + `${operatorInfo.positionOperator === "contains" || operatorInfo.positionOperator === "ends-with" ? "*" : ""}` + childValues + `${operatorInfo.positionOperator === "contains" || operatorInfo.positionOperator === "starts-with" ? "*" : ""}`;
        }
      }
      convertedString = addNodeToString(convertedString, statement);
      break;
    }
    case "JSONPath": {
      convertedString = addNodeToString(convertedString, tree.value);
      break;
    }
    case "Duration": {
      convertedString = addNodeToString(
        convertedString,
        `${tree.value}${tree.unit}`
      );
      break;
    }
    default:
      convertedString = addNodeToString(
        convertedString,
        getNestedValue({
          value: tree.value,
          textValue: tree.textValue,
          nodeType: tree.type
        })
      );
      break;
  }
  return convertedString;
}
function addNodeToString(sourceString, insertionString) {
  if (insertionString) {
    const firstElement = sourceString === "" || sourceString === "(";
    return `${sourceString}${firstElement ? "" : " "}${insertionString}`;
  }
  return sourceString;
}
