// @dynatrace/strato-components/filters/filter-field/utils/conversion/tree-precedence.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var tree_precedence_exports = {};
__export(tree_precedence_exports, {
  sortFilterFieldTree: () => sortFilterFieldTree
});
module.exports = __toCommonJS(tree_precedence_exports);
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
var import_tree_node_guards = require("../tree-node-guards.js");
var import_node_mutations = require("./lezer-tree-conversion/node-mutations.js");
function sortFilterFieldTree(tree, parserConfig, doc) {
  if (!tree) {
    return { tree, errors: [], isValidSearch: true };
  }
  const { tree: _cleanedTree, searchOperators } = groupByLogicalOperators({
    tree
  });
  let errors = [];
  let cleanedTree = _cleanedTree;
  let isValidSearch;
  if (parserConfig?.searchConversion === true) {
    ({
      tree: cleanedTree,
      errors,
      isValidSearch
    } = checkValidSearchOperators(cleanedTree, searchOperators, doc));
  }
  return {
    tree: cleanedTree,
    errors,
    isValidSearch: isValidSearch ?? false
  };
}
function groupByLogicalOperators(data, group) {
  const { tree, searchOperators = [] } = data;
  const sortedTree = group ?? {
    type: "Group",
    range: tree.range,
    logicalOperator: "AND",
    explicit: tree.explicit,
    explicitLogicalOperator: false,
    children: []
  };
  let prevLogicalOperator;
  let currentLogicalOperator;
  let nestedTree;
  if (tree.type === "Group" && (tree.children.length === 0 || tree.children[0].type === "Error")) {
    return {
      tree: {
        ...tree,
        logicalOperator: "AND",
        explicitLogicalOperator: false,
        children: tree.children
      },
      searchOperators
    };
  }
  for (const [index, child] of tree.children.entries()) {
    if (child.type === "Error" && !(0, import_tree_node_guards.isLogicalOperator)(child)) {
      sortedTree.children.push(child);
    }
    if (child.type === "Statement" || child.type === "Group" && child.explicit) {
      const prevNode = tree.children[index - 1];
      if (child.type === "Statement" && child.comparisonOperator?.type === "SearchOperator") {
        searchOperators.push(child.comparisonOperator);
      }
      prevLogicalOperator = currentLogicalOperator;
      currentLogicalOperator = (0, import_tree_node_guards.isLogicalOperator)(prevNode) ? (
        // DEPRECATION: [APPDEV-13773] BREAKING-CHANGE: Exclude logical operator node from public tree type.
        // Check how to best deal with the logical operator conversion here.
        prevNode.value.toUpperCase()
      ) : null;
      if (prevNode === void 0) {
        currentLogicalOperator = void 0;
      }
      const shouldCreateNestedTree = prevLogicalOperator !== void 0 && currentLogicalOperator !== void 0 && currentLogicalOperator !== prevLogicalOperator && // For an explicit logical operator, a nested tree has already been created when handling the logical operator.
      prevNode?.type !== "LogicalOperator";
      if (shouldCreateNestedTree) {
        const prevNode2 = sortedTree.children.pop();
        if (!prevNode2) {
          continue;
        }
        let nestedChildren;
        if (child.type === "Group") {
          ({ tree: nestedChildren } = groupByLogicalOperators({
            tree: child,
            searchOperators
          }));
        } else {
          nestedChildren = child;
        }
        const groupedChildren = [prevNode2, nestedChildren];
        nestedTree = {
          type: "Group",
          range: {
            from: prevNode2.range.from,
            to: groupedChildren[groupedChildren.length - 1]?.range.to
          },
          children: groupedChildren,
          logicalOperator: currentLogicalOperator ?? "AND",
          explicitLogicalOperator: currentLogicalOperator !== null && currentLogicalOperator !== void 0,
          explicit: false
        };
        sortedTree.children.push(nestedTree);
        if (sortedTree.range.to < nestedTree.range.to) {
          sortedTree.range.to = nestedTree.range.to;
        }
      } else {
        let nestedChildren;
        if (child.type === "Group") {
          ({ tree: nestedChildren } = groupByLogicalOperators(
            { tree: child, searchOperators },
            // Pass the created nested tree for implicit groups only to avoid circular structures.
            // The sorted tree returned at the end of the function is set to the 'group' parameter = nestedTree.
            // If the child is an explicit group and we would pass the nestedTree, it would be pushed into the
            // sortedTree of the previous recursion again when returning the current sortedTree.
            child.explicit ? void 0 : nestedTree
          ));
        } else {
          nestedChildren = child;
        }
        (nestedTree ?? sortedTree).children.push(nestedChildren);
        if (nestedTree) {
          nestedTree.range.to = child.range.to;
        }
        if (!sortedTree.explicit && sortedTree.range.to < child.range.to) {
          sortedTree.range.to = child.range.to;
        }
      }
    }
    if ((0, import_tree_node_guards.isLogicalOperator)(child)) {
      const convertedChild = child;
      prevLogicalOperator = currentLogicalOperator;
      currentLogicalOperator = convertedChild.value.toUpperCase();
      if (prevLogicalOperator !== void 0 && prevLogicalOperator !== currentLogicalOperator) {
        if (!nestedTree || currentLogicalOperator === "OR" && currentLogicalOperator !== sortedTree.logicalOperator && // Either of the trees has an explicit logical operator. The current logical operator does not match
        // the operator of the sorted tree. We need to move the existing sorted tree to a nested tree.
        (sortedTree.explicitLogicalOperator || nestedTree.explicitLogicalOperator || // Some deeper nested mixture of implicit and explicit logical and.
        // With a current logical OR operator, we need to move the existing sorted tree to a nested tree.
        sortedTree.logicalOperator === nestedTree.logicalOperator) || // The current logical operator is an explicit AND with a previous implicit AND and logical OR.
        // E.g.: b=2 OR c=3 d=4 AND e=5
        // In this case, we already have a nested tree, but still have to create a new one to group
        // The implicit and explicit logical AND while not losing the logical OR.
        sortedTree.logicalOperator === "OR" && currentLogicalOperator === "AND") {
          nestedTree = {
            type: "Group",
            logicalOperator: "AND",
            range: { from: -1, to: -1 },
            explicit: false,
            explicitLogicalOperator: currentLogicalOperator === "AND" || currentLogicalOperator === "OR" && sortedTree.explicitLogicalOperator,
            children: []
          };
          let abortGrouping = false;
          do {
            const previousNode = sortedTree.children.pop();
            if (!previousNode) {
              break;
            }
            nestedTree.children.push(previousNode);
            if (
              // Switching from logical and to logical or.
              sortedTree.explicitLogicalOperator && sortedTree.logicalOperator === "OR" || // Mixture of implicit and explicit logical and.
              !sortedTree.explicitLogicalOperator && currentLogicalOperator === "AND"
            ) {
              abortGrouping = true;
            }
          } while (!abortGrouping);
          nestedTree.children.reverse();
          nestedTree.range.from = nestedTree.children[0]?.range.from;
          nestedTree.range.to = nestedTree.children[nestedTree.children.length - 1]?.range.to;
          if (currentLogicalOperator === "AND" && nestedTree.range.to < child.range.to) {
            nestedTree.range.to = child.range.to;
          }
          if (!nestedTree.children[0]) {
            nestedTree = void 0;
          } else {
            sortedTree.children.push(nestedTree);
            if (currentLogicalOperator === "OR") {
              sortedTree.logicalOperator = currentLogicalOperator;
            }
          }
        }
        if (currentLogicalOperator === "OR" || nestedTree?.explicitLogicalOperator === false && currentLogicalOperator === "AND") {
          nestedTree = void 0;
        }
      } else {
        (nestedTree ?? sortedTree).explicitLogicalOperator = child.type !== "Error";
      }
      if (child.type === "Error") {
        (nestedTree ?? sortedTree).children.push(child);
      }
      if (!nestedTree) {
        if (!prevLogicalOperator) {
          sortedTree.logicalOperator = currentLogicalOperator;
        }
        if (currentLogicalOperator && child.type !== "Error") {
          sortedTree.explicitLogicalOperator = true;
        }
        if (sortedTree.range.to < child.range.to) {
          sortedTree.range.to = child.range.to;
        }
      }
    }
    if (child.type === "Error" && (child.value === "OR" || child.value === "AND")) {
      (nestedTree ?? sortedTree).explicitLogicalOperator = false;
    }
  }
  return { tree: sortedTree, searchOperators };
}
function checkValidSearchOperators(root, searchOperators, doc) {
  const errors = [];
  const hasStatementsOtherThanSearch = hasNonSearchStatements(root);
  const isSearchInNestedGroup = isSearchOperatorInsideGroup(root, true);
  const endsWithOR = typeof doc === "string" && doc.endsWith("OR ");
  const hasOr = root.logicalOperator === "OR" || endsWithOR;
  if (
    // For a group with an explicit or, we need at least two children that are not logical operators
    // for it to invalidate search operators.
    // E.g.:
    // ›foo = bar or‹ should still show the search suggestion, since the user may want to search for "or".
    // At this point, the "or" is already detected as logical operator (converted to an error node).
    searchOperators && (root.children.filter((child) => !(0, import_tree_node_guards.isLogicalOperator)(child)).length > 1 && hasStatementsOtherThanSearch && hasOr && searchOperators || isSearchInNestedGroup)
  ) {
    for (const node of searchOperators) {
      (0, import_node_mutations.convertToErrorNode)(node);
      errors.push({
        range: node.range,
        message: {
          type: "unsupportedSearchWithLogicalOr",
          details: `<span style="color: ${import_colors.default.Syntax.Operators}">AND</span>`
        }
      });
    }
  }
  const isValidSearch = errors.length === 0 && !(hasOr && hasStatementsOtherThanSearch) && !isSearchInNestedGroup;
  return { tree: root, errors, isValidSearch };
}
function hasNonSearchStatements(root) {
  return root.children.some((node) => {
    if (node.type === "Group") {
      return hasNonSearchStatements(node);
    }
    return (0, import_tree_node_guards.isFilterFieldStatementNode)(node) && node.comparisonOperator !== void 0 && node.comparisonOperator.type !== "SearchOperator";
  });
}
function isSearchOperatorInsideGroup(node, isRoot = true) {
  for (const child of node.children) {
    if (child.type === "Group") {
      if (isSearchOperatorInsideGroup(child, false)) {
        return true;
      }
    }
    if (!isRoot && (0, import_tree_node_guards.isFilterFieldStatementNode)(child) && child.comparisonOperator?.type === "Error" && (child.comparisonOperator.value === "* ~" || child.comparisonOperator.value === "*~")) {
      return true;
    }
  }
  return false;
}
