// @dynatrace/strato-components/filters/filter-field/utils/conversion/tree-conversion.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var tree_conversion_exports = {};
__export(tree_conversion_exports, {
  convertLezerTree: () => convertLezerTree,
  parserVersion: () => parserVersion
});
module.exports = __toCommonJS(tree_conversion_exports);
var import_tree_precedence = require("./tree-precedence.js");
var import_editor_config = require("../../editor-state-configs/editor-config.js");
var import_group = require("./lezer-tree-conversion/parent-node/group.js");
const parserVersion = "004";
function convertLezerTree({
  tree,
  view,
  numberParser,
  doc,
  internalValidatorMap,
  customTypes,
  parserConfig
}) {
  const errorsStateField = view?.state.field(import_editor_config.usedErrors, false);
  if (!doc || doc === "") {
    errorsStateField && (errorsStateField.errors = []);
    return { isValid: true, isValidSearch: true };
  }
  const { tree: convertedFlatTree, errors: initialConversionErrors } = convertFlatLezerTree({
    tree,
    numberParser,
    doc,
    internalValidatorMap,
    customTypes,
    parserConfig
  });
  const {
    tree: convertedTree,
    errors: sortedTreeErrors,
    isValidSearch
  } = (0, import_tree_precedence.sortFilterFieldTree)(convertedFlatTree, parserConfig, doc);
  const errors = [...initialConversionErrors, ...sortedTreeErrors];
  if (errorsStateField) {
    errors.sort((a, b) => a.range.from < b.range.from ? -1 : 1);
    errorsStateField.errors = errors;
  }
  if (convertedTree) {
    convertedTree.version = "007";
  }
  return {
    tree: convertedTree,
    isValid: errors.length === 0,
    isValidSearch: isValidSearch ?? sortedTreeErrors
  };
}
function convertFlatLezerTree({
  tree,
  numberParser,
  doc,
  internalValidatorMap,
  customTypes,
  parserConfig
}) {
  if (!doc || doc.toString() === "") {
    return { errors: [] };
  }
  let convertedTree = {
    type: "Group",
    children: [],
    range: { from: 0, to: tree.topNode.to },
    logicalOperator: "AND",
    explicitLogicalOperator: false,
    explicit: false
  };
  const errors = [];
  tree.iterate({
    from: tree.topNode.from,
    to: tree.topNode.to,
    enter: (_node) => {
      const node = _node.node;
      if (node.name === "Filter") {
        const { convertedNode, errors: groupErrors } = (0, import_group.convertFilterGroup)(
          node,
          doc,
          numberParser,
          internalValidatorMap,
          customTypes,
          parserConfig,
          "Filter"
        );
        convertedTree = convertedNode;
        errors.push(...groupErrors);
        return false;
      } else {
        console.log(
          `${doc.toString()} | Syntax node '${node.name}' cannot be converted. Please make sure to handle all nodes accordingly.`
        );
      }
    }
  });
  const filter = {
    tree: convertedTree,
    errors
  };
  return filter;
}
