// @dynatrace/strato-components/esm/charts/core/components/single-value-renderer/components/trend/Trend.js
import { jsx, jsxs } from "react/jsx-runtime";
import { isEmpty, isNil } from "lodash-es";
import { useContext, useLayoutEffect, useRef } from "react";
import { Icon } from "./components/Icon.js";
import {
  trendContainer,
  trendLabel,
  valueLabelContainer
} from "./Trend.sty.js";
import { defineTrendColors } from "./utils/define-trend-colors.js";
import { defineTrendDirection } from "./utils/define-trend-direction.js";
import { useId } from "../../../../../../core/hooks/useId.js";
import { useMergeRefs } from "../../../../../../core/hooks/useMergeRefs.js";
import { Flex } from "../../../../../../layouts/flex/Flex.js";
import { useFontSizeForContainer } from "../../../../hooks/useFontSizeForContainer.js";
import { useThrottledResizeObserver } from "../../../../hooks/useThrottledResizeObserver.js";
import { replaceSlashedZero } from "../../../../typographyStyles.sty.js";
import { getFormatter } from "../../../../utils/get-formatter.js";
import {
  FONT_HEIGHT_RATIO,
  TREND_FONT_CSS_TOKEN,
  TREND_LABEL_DEFAULT_COLOR
} from "../../constants.js";
import { SingleValueContext } from "../../contexts/single-value.context.js";
import { useGlobalFontSize } from "../../hooks/use-global-font-size.js";
const Trend = (props) => {
  const {
    direction: directionProp,
    showIcon = true,
    value: valueProp,
    label,
    inverseTrend = "none",
    colorsOverride,
    formatter: formatterProp,
    contrastColor
  } = props;
  const { alignment } = useContext(SingleValueContext);
  const direction = defineTrendDirection(directionProp, valueProp);
  const [iconColor, valueColor] = defineTrendColors(
    direction,
    inverseTrend,
    colorsOverride
  );
  const formatter = getFormatter(formatterProp);
  const formattedValue = !isNil(valueProp) ? formatter(valueProp) : "";
  const value = formattedValue.replace(" ", "");
  const ref = useRef(null);
  const { refCallback, width = 0, height = 0 } = useThrottledResizeObserver();
  const mergedRef = useMergeRefs([refCallback, ref]);
  const showTrendArrow = showIcon && (!isEmpty(value) || !isEmpty(label));
  const preferredFontSize = useFontSizeForContainer({
    container: { width, height: height * FONT_HEIGHT_RATIO },
    text: showIcon ? `00${value}0${label}` : `${value}0${label}`,
    element: ref.current
  });
  const id = useId();
  const { updateTrendFontSize } = useGlobalFontSize();
  useLayoutEffect(() => {
    if (height && width) {
      updateTrendFontSize?.(id, preferredFontSize);
    }
  }, [height, width, preferredFontSize]);
  return /* @__PURE__ */ jsxs(
    Flex,
    {
      className: trendContainer,
      style: {
        [TREND_FONT_CSS_TOKEN]: preferredFontSize
      },
      ref: mergedRef,
      "data-testid": "single-value-trend",
      alignItems: "center",
      justifyContent: alignment,
      gap: 0,
      children: [
        showTrendArrow && /* @__PURE__ */ jsx(Icon, { direction, color: contrastColor ?? iconColor }),
        /* @__PURE__ */ jsxs("div", { className: valueLabelContainer, children: [
          /* @__PURE__ */ jsx(
            "span",
            {
              className: replaceSlashedZero,
              style: {
                color: contrastColor ?? valueColor,
                paddingLeft: showTrendArrow ? "2ch" : "0"
              },
              children: value
            }
          ),
          /* @__PURE__ */ jsx(
            "span",
            {
              className: trendLabel,
              style: {
                color: contrastColor ?? TREND_LABEL_DEFAULT_COLOR
              },
              children: label
            }
          )
        ] })
      ]
    }
  );
};
Trend["displayName"] = "SingleValue.Trend";
export {
  Trend
};
