// @dynatrace/strato-components/esm/forms/password-input/PasswordInput.js
import { jsx } from "react/jsx-runtime";
import {
  forwardRef,
  useState,
  Children
} from "react";
import { useIntl } from "react-intl";
import { ViewIcon, ViewOffIcon } from "@dynatrace/strato-icons";
import {
  PasswordInputTooltip
} from "./PasswordInputTooltip.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useControllableState } from "../../core/hooks/useControllableState.js";
import { extractBehaviorTrackingProps } from "../../core/utils/extract-behavior-tracking-props.js";
import { isSlot } from "../../core/utils/is-slot.js";
import { mergeProps } from "../../core/utils/merge-props.js";
import { BaseInput } from "../_base-input/BaseInput.js";
import { getPlaceholderMessage } from "../_base-input/baseInputMessages.js";
const COMPONENT_NAME = "PasswordInput";
function getTooltipSlot(children) {
  const slots = {
    tooltip: null
  };
  for (const child of Children.toArray(children)) {
    if (isSlot(child, PasswordInputTooltip)) {
      slots.tooltip = child;
    } else if (child !== null) {
      console.warn("The used child is not supported by the PasswordInput.");
    }
  }
  return slots;
}
const PasswordInputRoot = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const intl = useIntl();
  const defaultPlaceholder = intl.formatMessage(
    getPlaceholderMessage("password")
  );
  const {
    onChange,
    name,
    readOnly = false,
    disabled,
    controlState,
    children,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "aria-disabled": ariaDisabledProp,
    className: consumerClassName,
    style: consumerStyle,
    defaultValue,
    value,
    placeholder = defaultPlaceholder,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
  const [hidden, setHidden] = useState(true);
  const [internalValue, setInternalValue] = useControllableState({
    defaultValue,
    value
  });
  const onInternalChange = (event) => {
    onChange?.(event.target.value, event);
    setInternalValue(event.target.value);
  };
  const { tooltip } = getTooltipSlot(children);
  const { showText, hideText, ...remainingTooltipProps } = tooltip?.props ?? {};
  const tooltipBehavioralTrackingProps = extractBehaviorTrackingProps(
    remainingTooltipProps
  );
  const showTooltipDefaultText = intl.formatMessage({
    defaultMessage: "Show password",
    id: "tgL6QEjmq+ArASFk",
    description: "Text in a tooltip shown when hovering the password input icon to show the password"
  });
  const hideTooltipDefaultText = intl.formatMessage({
    defaultMessage: "Hide password",
    id: "vMt3ZRHwBods5xzM",
    description: "Text in a tooltip shown when hovering the password input icon to hide the password"
  });
  const dataDtProperties = `value:${hidden ? "show" : "hide"}`;
  const showTooltipText = showText ?? showTooltipDefaultText;
  const hideTooltipText = hideText ?? hideTooltipDefaultText;
  return /* @__PURE__ */ jsx(
    BaseInput,
    {
      ...mergeProps(restProps, behavioralTrackingProps),
      value: internalValue,
      role: "textbox",
      controlState,
      name,
      type: hidden ? "password" : "text",
      onChange: onInternalChange,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ref: forwardedRef,
      className: consumerClassName,
      style: consumerStyle,
      readOnly,
      disabled,
      "aria-disabled": ariaDisabledProp,
      "data-dt-component": COMPONENT_NAME,
      placeholder,
      children: /* @__PURE__ */ jsx(BaseInput.Suffix, { children: /* @__PURE__ */ jsx(
        BaseInput.Button,
        {
          onClick: () => setHidden(!hidden),
          tooltipText: hidden ? showTooltipText : hideTooltipText,
          tooltipBehavioralTrackingProps,
          "aria-label": hidden ? showTooltipText : hideTooltipText,
          id: remainingTooltipProps.id,
          "aria-disabled": ariaDisabledProp,
          "data-dt-name": "ShowHidePassword",
          "data-dt-properties": dataDtProperties,
          disabled: readOnly ? false : disabled,
          children: hidden ? /* @__PURE__ */ jsx(ViewIcon, {}) : /* @__PURE__ */ jsx(ViewOffIcon, {})
        }
      ) })
    }
  );
});
const PasswordInput = Object.assign(PasswordInputRoot, {
  Tooltip: PasswordInputTooltip
});
PasswordInput.displayName = COMPONENT_NAME;
export {
  PasswordInput
};
