// @dynatrace/strato-components/charts/categorical-bar/components/interactions/CategoricalBarChartOverlays.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CategoricalBarChartOverlays_exports = {};
__export(CategoricalBarChartOverlays_exports, {
  CategoricalBarChartOverlays: () => CategoricalBarChartOverlays
});
module.exports = __toCommonJS(CategoricalBarChartOverlays_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_useCategoricalMouseEventHandlers = require("./hooks/useCategoricalMouseEventHandlers.js");
var import_CategoricalBarChartInspectOverlay = require("./overlays/CategoricalBarChartInspectOverlay.js");
var import_CategoricalBarChartOverlayRefContext = require("../../contexts/CategoricalBarChartOverlayRefContext.js");
var import_CategoricalBarChartToolbar = require("../toolbar/CategoricalBarChartToolbar.js");
const CategoricalBarChartOverlays = () => {
  const [shouldShowInspect, setShouldShowInspect] = (0, import_react.useState)(true);
  const overlayRef = (0, import_react.useRef)(null);
  const {
    handleMouseEnter,
    handleMouseMove,
    handleMouseLeave,
    handleFocus,
    handleBlur
  } = (0, import_useCategoricalMouseEventHandlers.useCategoricalMouseEventHandlers)();
  const handleToolbarMouseEnter = () => {
    setShouldShowInspect(false);
  };
  const handleToolbarMouseLeave = () => {
    setShouldShowInspect(true);
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      tabIndex: 0,
      ref: overlayRef,
      onFocus: handleFocus,
      onBlur: handleBlur,
      onMouseEnter: handleMouseEnter,
      onMouseMove: handleMouseMove,
      onMouseLeave: handleMouseLeave,
      style: { height: "100%" },
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_CategoricalBarChartOverlayRefContext.CategoricalBarChartOverlayRefContext.Provider, { value: overlayRef, children: [
        shouldShowInspect && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_CategoricalBarChartInspectOverlay.CategoricalBarChartInspectOverlay, {}),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_CategoricalBarChartToolbar.CategoricalBarChartToolbar,
          {
            onMouseEnter: handleToolbarMouseEnter,
            onMouseLeave: handleToolbarMouseLeave
          }
        )
      ] })
    }
  );
};
CategoricalBarChartOverlays["displayName"] = "CategoricalBarChartInteractions";
