// @dynatrace/strato-components/charts/tree-map/context/TreeMapAggregationDataContext.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TreeMapAggregationDataContext_exports = {};
__export(TreeMapAggregationDataContext_exports, {
  TreeMapAggregationDataContext: () => TreeMapAggregationDataContext,
  useTreeMapAggregationData: () => useTreeMapAggregationData
});
module.exports = __toCommonJS(TreeMapAggregationDataContext_exports);
var import_react = require("react");
var import_constants = require("../constants.js");
const defaultData = {
  children: [],
  name: "",
  value: 0,
  id: import_constants.DEFAULT_ROOT_NODE_NAME
};
const TreeMapAggregationDataContext = (0, import_react.createContext)({
  dataForRendering: defaultData,
  dataForHighlighting: defaultData
});
const useTreeMapAggregationData = () => (0, import_react.useContext)(TreeMapAggregationDataContext);
TreeMapAggregationDataContext["displayName"] = "TreeMapAggregationDataContext";
