// @dynatrace/strato-components/esm/forms/forms-core/components/form-field-messages/FormFieldMessageBaseItem.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import { CriticalIcon } from "@dynatrace/strato-icons";
import * as styles from "./FormFieldMessageItem.sty.js";
import { useBehavioralTrackingProps } from "../../../../core/hooks/useBehavioralTrackingProps.js";
import { Flex } from "../../../../layouts/flex/Flex.js";
import { useFormFieldMessagesContext } from "../../hooks/useFormFieldMessagesContext.js";
const InternalBaseItem = (props) => {
  const {
    children,
    variant = "hint",
    className: consumerClassName,
    style: consumerStyle,
    disabled: propDisabled = false,
    "data-testid": dataTestId,
    id: propId,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps] = useBehavioralTrackingProps(remainingProps);
  const { disabled: contextDisabled } = useFormFieldMessagesContext();
  const disabled = propDisabled || contextDisabled;
  return /* @__PURE__ */ jsxs(
    Flex,
    {
      className: clsx(
        styles.formFieldMessageItem({ variant, disabled }),
        consumerClassName
      ),
      style: consumerStyle,
      alignItems: "flex-start",
      gap: 4,
      "data-testid": dataTestId,
      id: propId,
      ...behavioralTrackingProps,
      children: [
        variant === "error" && /* @__PURE__ */ jsx(Flex, { children: /* @__PURE__ */ jsx(CriticalIcon, { size: "small" }) }),
        /* @__PURE__ */ jsx("div", { className: styles.formFieldMessageItemText, children })
      ]
    }
  );
};
InternalBaseItem.displayName = "FormFieldMessageItem";
export {
  InternalBaseItem
};
