// @dynatrace/strato-components/esm/forms/forms-core/components/form-field-messages/FormFieldMessages.js
import { jsx } from "react/jsx-runtime";
import { createElement } from "react";
import clsx from "clsx";
import { forwardRef } from "react";
import { FormFieldMessageItem } from "./FormFieldMessageItem.js";
import * as styles from "./FormFieldMessages.sty.js";
import { useBehavioralTrackingProps } from "../../../../core/hooks/useBehavioralTrackingProps.js";
import { Flex } from "../../../../layouts/flex/Flex.js";
import { useFormFieldContext } from "../../hooks/useFormFieldContext.js";
import { useFormFieldMessagesContext } from "../../hooks/useFormFieldMessagesContext.js";
const FormFieldMessagesRoot = /* @__PURE__ */ forwardRef((FormFieldMessageProps, forwardedRef) => {
  const {
    children,
    className: consumerClassName,
    style: consumerStyle,
    "data-dtrum-allow": dataDtrumAllow,
    "data-dtrum-mask": dataDtrumMask,
    "data-testid": dataTestId,
    ...remainingProps
  } = FormFieldMessageProps;
  const [behavioralTrackingProps] = useBehavioralTrackingProps(remainingProps);
  const { state } = useFormFieldMessagesContext();
  const { disabled, ariaDisabled } = useFormFieldContext();
  const errorMessages = state.formControlMessages.filter(
    (message) => message.variant === "error"
  );
  const internalHintMessages = state.formControlMessages.filter(
    (message) => message.variant === "hint"
  );
  const externalHintMessages = state.customMessages.filter(
    (message) => message.variant === "hint"
  );
  const hintMessages = [...internalHintMessages, ...externalHintMessages];
  const formMessages = state.showError ? errorMessages : internalHintMessages;
  const hideMessages = !state.showError && hintMessages.length === 0;
  let content = null;
  if (typeof children === "function") {
    content = children(formMessages);
  } else if (children !== void 0 && children !== null) {
    content = children;
  } else {
    content = formMessages.map((message) => /* @__PURE__ */ createElement(
      FormFieldMessages.Item,
      {
        disabled: disabled || ariaDisabled,
        ...message,
        key: message.id
      }
    ));
  }
  return /* @__PURE__ */ jsx(
    Flex,
    {
      ref: forwardedRef,
      flexDirection: "column",
      "aria-live": "polite",
      style: { ...consumerStyle, display: hideMessages ? "none" : "flex" },
      className: clsx(consumerClassName, styles.formFieldMessages),
      "aria-hidden": hideMessages,
      "data-dtrum-allow": dataDtrumAllow,
      "data-dtrum-mask": dataDtrumMask,
      "data-testid": dataTestId,
      ...behavioralTrackingProps,
      children: content
    }
  );
});
FormFieldMessagesRoot.displayName = "FormFieldMessages";
const FormFieldMessages = Object.assign(FormFieldMessagesRoot, {
  Item: FormFieldMessageItem
});
export {
  FormFieldMessages
};
