// @dynatrace/strato-components/filters/segment-selector/parts/selectionBar/SegmentSelectorTrigger.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SegmentSelectorTrigger_exports = {};
__export(SegmentSelectorTrigger_exports, {
  SegmentSelectorTrigger: () => SegmentSelectorTrigger
});
module.exports = __toCommonJS(SegmentSelectorTrigger_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_strato_icons = require("@dynatrace/strato-icons");
var styles = __toESM(require("./SegmentSelectorTrigger.sty.js"));
var import_Button = require("../../../../buttons/button/Button.js");
var import_merge_props = require("../../../../core/utils/merge-props.js");
var import_TextEllipsis = require("../../../../typography/text-ellipsis/TextEllipsis.js");
var import_useTriggerContext = require("../../hooks/useTriggerContext.js");
const SegmentSelectorTrigger = (props) => {
  const {
    textEllipsisRef,
    className: consumerClassName,
    ...remainingProps
  } = props;
  const { triggerProps, displayValue, isLoading, buttonRef, showError } = (0, import_useTriggerContext.useSegmentsTriggerContext)();
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_Button.Button,
    {
      ref: buttonRef,
      loading: isLoading,
      variant: "emphasized",
      textAlign: "start",
      color: showError ? "critical" : "neutral",
      className: (0, import_clsx.default)(styles.trigger, consumerClassName),
      ...(0, import_merge_props.mergeProps)(triggerProps, remainingProps),
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ContainerIcon, {}) }),
        displayValue && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TextEllipsis.TextEllipsis, { ref: textEllipsisRef, children: displayValue }),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Suffix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ChevronDownSmallIcon, {}) })
      ]
    }
  );
};
SegmentSelectorTrigger.displayName = "SegmentSelectorTrigger";
