// @dynatrace/strato-components/filters/segment-selector/parts/selectionBar/SegmentSelect.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SegmentSelect_exports = {};
__export(SegmentSelect_exports, {
  SegmentSelect: () => SegmentSelect
});
module.exports = __toCommonJS(SegmentSelect_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_intl = require("react-intl");
var styles = __toESM(require("./SegmentSelect.sty.js"));
var import_useMergeRefs = require("../../../../core/hooks/useMergeRefs.js");
var import_is_node_environment = require("../../../../core/utils/is-node-environment.js");
var import_Select = require("../../../../forms/select/Select.js");
var import_Text = require("../../../../typography/text/Text.js");
var import_SegmentsCache = require("../../contexts/SegmentsCache/SegmentsCache.js");
var import_SegmentsUIContext = require("../../contexts/SegmentsUI/SegmentsUIContext.js");
var import_segmentMessages = require("../../segmentMessages.js");
function getSafeSegmentName(segment, fallback) {
  return segment?.name && segment.name !== "" ? segment.name : fallback;
}
const SegmentSelect = (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    availableSegments,
    readOnly,
    triggerClassName,
    overlayWidth,
    segment: cachedSegment,
    autoFocus,
    isPlaceholder,
    id
  } = props;
  const intl = (0, import_react_intl.useIntl)();
  const segmentSelectRef = (0, import_react.useRef)(null);
  const mergedRef = (0, import_useMergeRefs.useMergeRefs)([forwardedRef, segmentSelectRef]);
  const variableSelectRef = (0, import_react.useRef)(null);
  const { updateUISegment, removeUISegment } = (0, import_SegmentsUIContext.useSegmentsUIContext)();
  const { getSegment } = (0, import_SegmentsCache.useSegmentsCache)();
  const { uiSegments } = (0, import_react.useContext)(import_SegmentsUIContext.UISegmentsContext);
  const refIsDefined = segmentSelectRef.current !== null;
  (0, import_react.useEffect)(() => {
    import_SegmentsCache.segmentsClient.hasAccessRights && segmentSelectRef.current?.validate();
    if (uiSegments.length === 1 || autoFocus) {
      segmentSelectRef.current?.focus?.();
      return;
    }
    if (isPlaceholder && !autoFocus) {
      segmentSelectRef.current?.open();
    }
  }, [refIsDefined]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_Select.Select,
    {
      ref: mergedRef,
      className: styles.segmentsSelectCSS,
      value: id,
      id,
      required: !isPlaceholder,
      clearable: false,
      readOnly: readOnly || cachedSegment?.temporarilyApplied,
      onOpenChange: (isOpen) => {
        if (isOpen) {
          return;
        }
        variableSelectRef.current?.open();
      },
      onChange: (segmentUid) => {
        if (segmentUid) {
          getSegment(segmentUid).then((newSegment) => {
            updateUISegment(
              { id },
              {
                ...newSegment.segment,
                variables: !newSegment.segment.variables ? void 0 : []
              }
            );
          }).catch((error) => {
            if ((0, import_is_node_environment.isDev)()) {
              console.log(
                `SegmentSelector: fetching the segment with id ${id} failed.`,
                error
              );
            }
            removeUISegment(id);
          });
        }
      },
      "data-dt-name": "segment-select",
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_Select.Select.Trigger,
          {
            width: "full",
            className: (0, import_clsx.default)(triggerClassName, {
              [styles.segmentsSelectTriggerCSS]: cachedSegment?.segment?.variables
            }),
            placeholder: intl.formatMessage(
              import_segmentMessages.SEGMENTSELECTOR_MESSAGES.segmentsTriggerPlaceholder
            )
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Select.Select.Filter, {}),
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Select.Select.Content, { showSelectedOptionsFirst: true, width: overlayWidth, children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Select.Select.EmptyState, { children: availableSegments.length === 0 ? intl.formatMessage(
            import_segmentMessages.SEGMENTSELECTOR_MESSAGES.segmentsSelectEmptyState
          ) : intl.formatMessage(
            import_segmentMessages.SEGMENTSELECTOR_MESSAGES.segmentsSelectEmptyStateSearch
          ) }),
          (id && !isPlaceholder || cachedSegment?.segment) && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_Select.Select.Option,
            {
              value: id,
              textValue: getSafeSegmentName(cachedSegment?.segment, id),
              children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { children: getSafeSegmentName(cachedSegment?.segment, id) })
            },
            id
          ),
          availableSegments.map((segmentOption) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_Select.Select.Option,
            {
              value: segmentOption.id,
              textValue: getSafeSegmentName(segmentOption, segmentOption.id),
              children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { children: getSafeSegmentName(segmentOption, segmentOption.id) })
            },
            segmentOption.id
          ))
        ] })
      ]
    }
  );
});
