// @dynatrace/strato-components/filters/segment-selector/parts/selectionBar/SegmentSelect.sty.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SegmentSelect_css_exports = {};
__export(SegmentSelect_css_exports, {
  clearButtonCSS: () => clearButtonCSS,
  noBorderRadiusRight: () => noBorderRadiusRight,
  segmentsSelectCSS: () => segmentsSelectCSS,
  segmentsSelectGridCSS: () => segmentsSelectGridCSS,
  segmentsSelectTriggerCSS: () => segmentsSelectTriggerCSS,
  variablesSelectCSS: () => variablesSelectCSS,
  variablesSelectTriggerCSS: () => variablesSelectTriggerCSS
});
module.exports = __toCommonJS(SegmentSelect_css_exports);
var import_AppRoot_css_ts_vanilla = require("../../../../core/components/app-root/AppRoot.css");
var import_SegmentSelect_css_ts_vanilla = require("./SegmentSelect.css");
var import_createRuntimeFn = require("@vanilla-extract/recipes/createRuntimeFn");
var clearButtonCSS = "_1fuxi5m5-3-6-0";
var noBorderRadiusRight = "_1fuxi5m2-3-6-0";
var segmentsSelectCSS = "_1fuxi5m0-3-6-0";
var segmentsSelectGridCSS = (0, import_createRuntimeFn.createRuntimeFn)({ defaultClassName: "_1fuxi5m6-3-6-0", variantClassNames: { layout: { full: "_1fuxi5m7-3-6-0", "full-readonly": "_1fuxi5m8-3-6-0", "no-variables": "_1fuxi5m9-3-6-0", "no-variables-readonly": "_1fuxi5ma-3-6-0" } }, defaultVariants: {}, compoundVariants: [] });
var segmentsSelectTriggerCSS = "_1fuxi5m1-3-6-0";
var variablesSelectCSS = "_1fuxi5m3-3-6-0";
var variablesSelectTriggerCSS = "_1fuxi5m4-3-6-0";
