// @dynatrace/strato-components/filters/segment-selector/parts/selectionBar/SelectionBar.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SelectionBar_exports = {};
__export(SelectionBar_exports, {
  ErrorMessages: () => ErrorMessages,
  SelectionBar: () => SelectionBar
});
module.exports = __toCommonJS(SelectionBar_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_query = require("@tanstack/react-query");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_SegmentSelect_css = require("./SegmentSelect.sty.js");
var import_SegmentSelect = require("./SegmentSelect.js");
var import_ValueSelect = require("./ValueSelect.js");
var import_Button = require("../../../../buttons/button/Button.js");
var import_useOverlayStacking = require("../../../../core/hooks/useOverlayStacking.js");
var import_FormFieldMessages = require("../../../../forms/forms-core/components/form-field-messages/FormFieldMessages.js");
var import_useFormFieldMessagesContext = require("../../../../forms/forms-core/hooks/useFormFieldMessagesContext.js");
var import_Grid = require("../../../../layouts/grid/Grid.js");
var import_Tooltip = require("../../../../overlays/tooltip/Tooltip.js");
var import_SegmentsCache = require("../../contexts/SegmentsCache/SegmentsCache.js");
var import_SegmentsUIContext = require("../../contexts/SegmentsUI/SegmentsUIContext.js");
var import_segmentValidationHelper = require("../../contexts/SegmentsUI/segmentValidationHelper.js");
var import_segmentMessages = require("../../segmentMessages.js");
var listStyles = __toESM(require("../../SegmentSelector.sty.js"));
function getVariables(segment) {
  if (segment?.variables && segment.variables.length > 0) {
    return segment.variables[0].values;
  }
  return void 0;
}
function getLayout(hasVariables, hasClearButton) {
  if (!hasVariables && !hasClearButton) {
    return "no-variables-readonly";
  }
  if (!hasVariables && hasClearButton) {
    return "no-variables";
  }
  if (hasVariables && !hasClearButton) {
    return "full-readonly";
  }
  return "full";
}
function getErrorMessageFromKey(errorKey) {
  switch (errorKey) {
    case "unavailable":
      return import_segmentMessages.SEGMENTSELECTOR_MESSAGES.unavailableSegmentError;
    case "needs-update":
      return import_segmentMessages.SEGMENTSELECTOR_MESSAGES.removedVariablesError;
    case "needs-variables":
      return import_segmentMessages.SEGMENTSELECTOR_MESSAGES.valueMissingError;
    default:
      return null;
  }
}
const ErrorMessages = ({ uiSegment }) => {
  const {
    data: errorKey,
    isLoading,
    error
  } = (0, import_react_query.useQuery)(
    {
      queryKey: ["errorKey", uiSegment],
      queryFn: () => (0, import_segmentValidationHelper.getErrorKey)(uiSegment)
    },
    import_SegmentsCache.segmentsClient.queryClient
  );
  const { arrayIdentifierErrorIds } = (0, import_react.useContext)(import_SegmentsUIContext.UISegmentsContext);
  const hasArrayIdentifierError = arrayIdentifierErrorIds?.includes(
    uiSegment.id
  );
  if (import_SegmentsCache.segmentsClient.hasAccessRights === false || isLoading || error) {
    return null;
  }
  const errorMessage = getErrorMessageFromKey(errorKey);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_FormFieldMessages.FormFieldMessages, { children: (messages) => {
    if (hasArrayIdentifierError) {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_FormFieldMessages.FormFieldMessages.Item,
        {
          validityStateKey: "customError",
          variant: "error",
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_react_intl.FormattedMessage,
            {
              ...import_segmentMessages.SEGMENTSELECTOR_MESSAGES.arrayIdentifierError
            }
          )
        },
        "array-identifier"
      );
    }
    if (errorKey) {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_FormFieldMessages.FormFieldMessages.Item,
        {
          validityStateKey: "customError",
          variant: "error",
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_intl.FormattedMessage, { ...errorMessage })
        },
        errorKey
      );
    }
    return messages.map((props) => {
      if (props.controlId !== uiSegment.id && props.validityStateKey === "valueMissing") {
        return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_FormFieldMessages.FormFieldMessages.Item,
          {
            ...props,
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_react_intl.FormattedMessage,
              {
                ...import_segmentMessages.SEGMENTSELECTOR_MESSAGES.valueMissingError
              }
            )
          },
          `${props.controlId}-valueMissing`
        );
      }
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_FormFieldMessages.FormFieldMessages.Item,
        {
          ...props
        },
        `${props.controlId}-${props.validityStateKey}`
      );
    });
  } });
};
const SelectionBar = (props) => {
  const {
    availableSegments,
    clearable,
    readOnly,
    autoFocus,
    uiSegment,
    returnFocusTo,
    segment: propSegment
  } = props;
  const intl = (0, import_react_intl.useIntl)();
  const gridRef = (0, import_react.useRef)(null);
  const isPlaceholder = uiSegment.isPlaceholder ?? false;
  const id = uiSegment.id;
  const variables = getVariables(uiSegment);
  const { removeUISegment } = (0, import_SegmentsUIContext.useSegmentsUIContext)();
  const { state } = (0, import_useFormFieldMessagesContext.useFormFieldMessagesContext)();
  const { getSegment } = (0, import_SegmentsCache.useSegmentsCache)();
  const { data: fetchedSegment } = (0, import_react_query.useQuery)(
    {
      queryKey: ["getSegment", id],
      queryFn: () => getSegment(id),
      enabled: import_SegmentsCache.segmentsClient.hasAccessRights !== false && !isPlaceholder && propSegment === void 0,
      retryOnMount: false
    },
    import_SegmentsCache.segmentsClient.queryClient
  );
  const segment = propSegment ?? fetchedSegment;
  const showDeleteButton = clearable && !readOnly && import_SegmentsCache.segmentsClient.hasAccessRights !== false;
  const stackingContext = (0, import_useOverlayStacking.useOverlayStacking)(void 0, "overlay");
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("li", { className: listStyles.segmentSelectorListItem, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Tooltip.Tooltip,
      {
        text: intl.formatMessage(
          import_segmentMessages.SEGMENTSELECTOR_MESSAGES.unlistedSegmentTooltip
        ),
        style: { zIndex: Number(stackingContext.style.zIndex) + 1 },
        disabled: !segment?.temporarilyApplied,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
          import_Grid.Grid,
          {
            ref: gridRef,
            className: (0, import_SegmentSelect_css.segmentsSelectGridCSS)({
              layout: getLayout(
                !!segment?.segment?.variables,
                !!showDeleteButton
              )
            }),
            children: [
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_SegmentSelect.SegmentSelect,
                {
                  readOnly,
                  availableSegments,
                  segment,
                  id,
                  overlayWidth: `${gridRef.current?.clientWidth}px`,
                  triggerClassName: (0, import_clsx.default)({
                    [import_SegmentSelect_css.noBorderRadiusRight]: showDeleteButton || segment?.segment?.variables
                  }),
                  isPlaceholder,
                  autoFocus
                }
              ),
              segment?.segment?.variables && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_ValueSelect.ValueSelect,
                {
                  readOnly,
                  segment,
                  variables,
                  triggerClassName: (0, import_clsx.default)({
                    [import_SegmentSelect_css.noBorderRadiusRight]: clearable && !readOnly
                  })
                }
              ),
              showDeleteButton && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_Button.Button,
                {
                  onClick: () => {
                    removeUISegment(id);
                    returnFocusTo.current?.focus();
                  },
                  color: state.showError ? "critical" : "neutral",
                  variant: "emphasized",
                  className: import_SegmentSelect_css.clearButtonCSS,
                  "aria-label": intl.formatMessage(
                    import_segmentMessages.SEGMENTSELECTOR_MESSAGES.removeSegmentButton
                  ),
                  "data-dt-name": "Remove",
                  children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.XmarkIcon, {}) })
                }
              )
            ]
          }
        )
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(ErrorMessages, { uiSegment })
  ] });
};
