// @dynatrace/strato-components/filters/segment-selector/parts/selectionBar/ValueSelect.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ValueSelect_exports = {};
__export(ValueSelect_exports, {
  ValueSelect: () => ValueSelect
});
module.exports = __toCommonJS(ValueSelect_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
var import_react_hooks = require("@dynatrace-sdk/react-hooks");
var styles = __toESM(require("./SegmentSelect.sty.js"));
var import_StratoLoggerContext = require("../../../../core/sdk/StratoLoggerContext.js");
var import_useFormFieldMessagesContext = require("../../../../forms/forms-core/hooks/useFormFieldMessagesContext.js");
var import_formfield_messages_state_reducer = require("../../../../forms/forms-core/state/formfield-messages-state-reducer.js");
var import_Select = require("../../../../forms/select/Select.js");
var import_Grid = require("../../../../layouts/grid/Grid.js");
var import_Link = require("../../../../typography/link/Link.js");
var import_Text = require("../../../../typography/text/Text.js");
var import_TextEllipsis = require("../../../../typography/text-ellipsis/TextEllipsis.js");
var import_SegmentsUIContext = require("../../contexts/SegmentsUI/SegmentsUIContext.js");
var import_segmentMessages = require("../../segmentMessages.js");
var import_utils = require("../../utils.js");
var import_segmentManagementHref = require("../segmentManagementHref.js");
const dtClientContext = JSON.stringify({
  source: "SegmentSelector",
  package: "@dynatrace/strato-components"
});
const ValueSelect = (props) => {
  const { triggerClassName, readOnly, segment, variables = [] } = props;
  const variableSelectRef = (0, import_react.useRef)(null);
  const { focusedSegmentId, isTouched } = (0, import_react.useContext)(import_SegmentsUIContext.UISegmentsContext);
  const { updateUISegment, setArrayIdentifierError } = (0, import_SegmentsUIContext.useSegmentsUIContext)();
  const { dispatch: messagesDispatch } = (0, import_useFormFieldMessagesContext.useFormFieldMessagesContext)();
  const intl = (0, import_react_intl.useIntl)();
  const stratoLogger = (0, import_react.useContext)(import_StratoLoggerContext.StratoLoggerContext);
  const query = segment?.segment?.variables !== void 0 && segment?.segment?.variables?.type === "query" ? segment.segment.variables.value : "";
  const {
    data,
    isLoading: isDqlLoading,
    forceRefetch
  } = (0, import_react_hooks.useDql)({
    query,
    maxResultRecords: 1e4,
    maxResultBytes: 5242880,
    // 5 MB
    dtClientContext
  });
  (0, import_react.useEffect)(() => {
    if (!isDqlLoading && data?.records?.length === 0) {
      stratoLogger?.info(
        "Failed to retrieve segment variables: query returned no records.",
        null,
        {
          query: query.split("|")[0],
          // log query until first pipe to avoid reporting sensitive data.
          id: segment?.segment?.id
        }
      );
    }
  }, [query, isDqlLoading]);
  const unavailableVariables = (0, import_react.useMemo)(() => {
    if (data) {
      const firstKey = Object.keys(data.records?.[0] ?? {})[0];
      const dataRecordIds = data.records.map((record) => record?.[firstKey]);
      return variables?.filter((variable) => !dataRecordIds.includes(variable));
    }
  }, [data, variables]);
  const availableVariables = (0, import_react.useMemo)(() => {
    if (data?.records) {
      let hasArrayIdentifier = false;
      const recordsArray = data?.records.map((singleRow) => {
        const firstKey = singleRow ? Object.keys(singleRow)[0] : null;
        if (!singleRow || !firstKey) {
          return null;
        }
        if ((0, import_lodash_es.isArray)(singleRow[firstKey])) {
          hasArrayIdentifier = true;
        }
        return `${singleRow[firstKey]}`;
      }).filter((value) => value !== "null" && value !== null);
      queueMicrotask(() => {
        if (hasArrayIdentifier) {
          messagesDispatch?.({
            type: import_formfield_messages_state_reducer.FormFieldMessagesAction.ADDMESSAGE,
            payload: {
              variant: "error",
              messageType: "formcontrol",
              id: "array-identifier",
              validityStateKey: "customError",
              message: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_react_intl.FormattedMessage,
                {
                  ...import_segmentMessages.SEGMENTSELECTOR_MESSAGES.arrayIdentifierError
                }
              )
            }
          });
        }
        setArrayIdentifierError(hasArrayIdentifier, segment.segment.id);
      });
      return Array.from(new Set(recordsArray));
    }
  }, [data]);
  const shouldOpen = focusedSegmentId === segment.segment.id && variables.length === 0 && isTouched;
  (0, import_react.useEffect)(() => {
    shouldOpen && variableSelectRef.current?.open();
  }, [shouldOpen]);
  (0, import_react.useEffect)(() => {
    !isTouched && variableSelectRef.current?.validate();
  }, [isTouched, JSON.stringify(variables)]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_Select.Select,
    {
      ref: variableSelectRef,
      multiple: true,
      className: styles.variablesSelectCSS,
      required: true,
      readOnly: readOnly || segment?.temporarilyApplied,
      clearable: false,
      onOpenChange: (isOpen) => {
        if (query.length > 0 && isOpen) {
          forceRefetch().catch(() => void 0);
        }
      },
      value: variables,
      onChange: (newVariables) => {
        if (!newVariables || newVariables.length === 0) {
          updateUISegment(
            {
              ...segment.segment,
              variables: []
            },
            {
              ...segment.segment,
              variables: []
            }
          );
          return;
        }
        const firstKey = Object.keys(data?.records[0] ?? {})[0];
        const dataEnrichedVariables = newVariables.map((variable) => {
          return data?.records.find((singleRow) => {
            if ((0, import_lodash_es.isArray)(singleRow?.[firstKey])) {
              const values = singleRow?.[firstKey].includes(variable);
              return values;
            }
            return singleRow && singleRow[firstKey] === variable;
          }) ?? null;
        }).filter((value) => value !== null);
        const segmentVariables = (0, import_utils.convertQueryRecordToStoreSegmentVariables)(
          dataEnrichedVariables
        );
        if (segmentVariables) {
          updateUISegment(
            {
              ...segment.segment,
              variables: []
            },
            {
              ...segment.segment,
              variables: segmentVariables
            }
          );
        }
      },
      "data-dt-name": "segment-variable-select",
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_Select.Select.Trigger,
          {
            className: (0, import_clsx.default)(styles.variablesSelectTriggerCSS, triggerClassName),
            placeholder: intl.formatMessage(
              import_segmentMessages.SEGMENTSELECTOR_MESSAGES.segmentsVariableTriggerPlaceholder
            ),
            children: variables?.length > 0 && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Select.Select.DisplayValue, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TextEllipsis.TextEllipsis, { children: variables.map((variable) => (0, import_utils.getReadableValue)(variable, intl)).join(", ") }) })
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Select.Select.Filter, {}),
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
          import_Select.Select.Content,
          {
            loading: isDqlLoading,
            showSelectedOptionsFirst: true,
            width: "trigger",
            children: [
              !data && variables?.map((variable) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_Select.Select.Option,
                {
                  value: variable,
                  textValue: (0, import_utils.getReadableValue)(variable, intl) ?? void 0,
                  children: (0, import_utils.getReadableValue)(variable, intl)
                },
                variable
              )),
              unavailableVariables?.map((unlistedVariable) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_Select.Select.Option,
                {
                  value: unlistedVariable,
                  textValue: (0, import_utils.getReadableValue)(unlistedVariable, intl) ?? void 0,
                  children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
                    import_Grid.Grid,
                    {
                      gridTemplateColumns: "auto auto",
                      width: "100%",
                      alignItems: "baseline",
                      children: [
                        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { children: (0, import_utils.getReadableValue)(unlistedVariable, intl) }),
                        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                          import_Text.Text,
                          {
                            textStyle: "small",
                            style: {
                              color: import_colors.default.Text.Neutral.Subdued,
                              textAlign: "right",
                              textWrap: "nowrap"
                            },
                            children: intl.formatMessage(import_segmentMessages.SEGMENTSELECTOR_MESSAGES.unavailable)
                          }
                        )
                      ]
                    }
                  )
                },
                unlistedVariable
              )),
              availableVariables?.map((variable) => /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
                import_Select.Select.Option,
                {
                  value: variable,
                  textValue: (0, import_utils.getReadableValue)(variable, intl) ?? void 0,
                  children: [
                    " ",
                    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { children: (0, import_utils.getReadableValue)(variable, intl) })
                  ]
                },
                variable
              )),
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Select.Select.EmptyState, { children: data?.records.length ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_react_intl.FormattedMessage,
                {
                  id: "nIw/xeiGNJsj+a+5",
                  defaultMessage: "No matching values for this segment found.",
                  description: "Hint shown when searching for a segment variable that isn't in the list."
                }
              ) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_react_intl.FormattedMessage,
                {
                  defaultMessage: "No values for this segment available. You can manage this segment in {link}.",
                  id: "72DlIwYKB9UTEzlA",
                  description: "Empty state message if no variables for the segment are available.",
                  values: {
                    link: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                      import_Link.Link,
                      {
                        href: (0, import_segmentManagementHref.getSegmentHref)(segment.segment.id),
                        target: "_blank",
                        rel: "noopener noreferrer",
                        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                          import_react_intl.FormattedMessage,
                          {
                            defaultMessage: "settings",
                            id: "u55UYpLvXi//xGEn",
                            description: "Link text to segments app in empty state message if no variables for the segment are available."
                          }
                        )
                      }
                    )
                  }
                }
              ) })
            ]
          }
        )
      ]
    }
  );
};
