// @dynatrace/strato-components/filters/filter/types/filter-state.types.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var filter_state_types_exports = {};
__export(filter_state_types_exports, {
  FilterBarActions: () => FilterBarActions,
  PinnedStateValues: () => PinnedStateValues
});
module.exports = __toCommonJS(filter_state_types_exports);
const PinnedStateValues = [
  /** The item is not displayed in the filter bar, but is available in the add-filter select and can change to a "pinned-optional" item at any time. */
  "optional",
  /** The item is currently displayed in the filter bar, but is available in the add-filter select and can change to an "optional" item at any time. */
  "pinned-optional",
  /** The item is always in the filter bar and not controllable with the add-filter select. Changing from "pinned" to any other state is also not recommended since that might break UX patterns.*/
  "pinned"
];
var FilterBarActions = /* @__PURE__ */ ((FilterBarActions2) => {
  FilterBarActions2["INITIALIZE"] = "initialize";
  FilterBarActions2["CHANGE"] = "change";
  FilterBarActions2["REMOVE"] = "remove";
  FilterBarActions2["PIN"] = "pin";
  FilterBarActions2["UNPIN"] = "unpin";
  FilterBarActions2["CHANGEPINNEDSTATE"] = "changepinnedstate";
  return FilterBarActions2;
})(FilterBarActions || {});
