// @dynatrace/strato-components/core/components/app-root/AppRoot.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var AppRoot_exports = {};
__export(AppRoot_exports, {
  AppRoot: () => AppRoot
});
module.exports = __toCommonJS(AppRoot_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_wicg_inert = require("wicg-inert");
var import_user_preferences = require("@dynatrace-sdk/user-preferences");
var styles = __toESM(require("./AppRoot.sty.js"));
var import_useCurrentTheme = require("../../hooks/useCurrentTheme.js");
var import_FocusProvider = require("../../providers/FocusProvider.js");
var import_is_visual_refresh_enabled = require("../../utils/is-visual-refresh-enabled.js");
var import_modality = require("../../utils/modality.js");
(0, import_modality.initModalityDetection)();
const COMPONENT_NAME = "AppRoot";
function getBaseHref() {
  const baseElement = document.querySelector("base");
  const href = baseElement?.href ?? "/";
  return href.endsWith("/") ? href : `${href}/`;
}
const defaultLanguage = "en";
const AppRoot = /* @__PURE__ */ (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const {
      style: consumerStyle,
      className: consumerClassName,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      translationRootPath = `${getBaseHref()}lang/`
    } = props;
    const [messages, setMessages] = (0, import_react.useState)({});
    const { children } = props;
    const theme = (0, import_useCurrentTheme.useCurrentTheme)();
    const language = (0, import_user_preferences.getLanguage)();
    let timezone = (0, import_user_preferences.getTimezone)();
    try {
      new Intl.DateTimeFormat(language, { timeZone: timezone });
    } catch {
      timezone = Intl.DateTimeFormat().resolvedOptions().timeZone;
    }
    (0, import_react.useLayoutEffect)(() => {
      if (document !== void 0) {
        document.documentElement.setAttribute("data-theme", theme);
      }
    }, [theme]);
    (0, import_react.useLayoutEffect)(() => {
      if (document !== void 0 && (0, import_is_visual_refresh_enabled.isVisualRefreshEnabled)()) {
        document.documentElement.classList.add(styles.dtVisualRefreshEnabled);
      }
    }, []);
    (0, import_react.useEffect)(() => {
      if (language === "en") {
        return;
      }
      fetch(`${translationRootPath}${language}.json`).then((res) => {
        return res.json();
      }).then((receivedMessages) => {
        setMessages(receivedMessages);
      }).catch(() => {
        setMessages((prev) => (0, import_lodash_es.isEmpty)(prev) ? prev : {});
      });
    }, [language]);
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "div",
      {
        ref: forwardedRef,
        "data-testid": dataTestId,
        className: consumerClassName,
        style: consumerStyle,
        "data-theme": theme,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        "data-dt-component": COMPONENT_NAME,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_react_intl.IntlProvider,
          {
            locale: language,
            timeZone: timezone,
            messages,
            defaultLocale: defaultLanguage,
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_FocusProvider.FocusProvider, { children })
          }
        )
      }
    );
  }
);
AppRoot.displayName = COMPONENT_NAME;
