// @dynatrace/strato-components/charts/core/components/tooltip/InteractiveChartTooltip/headers/TimebasedTooltipHeader.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TimebasedTooltipHeader_exports = {};
__export(TimebasedTooltipHeader_exports, {
  TimebasedTooltipHeader: () => TimebasedTooltipHeader
});
module.exports = __toCommonJS(TimebasedTooltipHeader_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_date_fns = require("date-fns");
var import_units = require("@dynatrace-sdk/units");
var import_TimeResolution = require("./subheaders/TimeResolution.js");
var import_FormattedTimeframe = require("../../utils/FormattedTimeframe.js");
var import_InteractiveChartTooltip_styles = require("../components/InteractiveChartTooltip.styles.js");
var import_SeriesBasedContent = require("../contents/SeriesBasedContent.js");
const TimebasedTooltipHeader = ({
  timeframe: [from, timeframeTo],
  item,
  seriesDisplayMode = "single-line"
}) => {
  const to = timeframeTo !== void 0 && (0, import_date_fns.isBefore)(from, timeframeTo) ? timeframeTo : void 0;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_InteractiveChartTooltip_styles.StyledChartTooltipHeader, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_InteractiveChartTooltip_styles.StyledTitleContainer, { gap: 8, children: [
      to && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TimeResolution.TimeResolution, { from, to }),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_InteractiveChartTooltip_styles.StyledFormattedTimeframe, { children: to !== void 0 ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_FormattedTimeframe.FormattedTimeframe,
        {
          "data-testid": "formatted-timeframe",
          from,
          to
        }
      ) : (0, import_units.formatDate)(from, {
        day: "2-digit",
        month: "short",
        year: from.getFullYear() === (/* @__PURE__ */ new Date()).getFullYear() ? void 0 : "numeric",
        hour: "2-digit",
        minute: "2-digit",
        second: from.getSeconds() === 0 ? void 0 : "numeric"
      }) })
    ] }),
    item && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_SeriesBasedContent.SeriesBasedContent,
      {
        item,
        highlighted: true,
        seriesDisplayMode
      }
    )
  ] });
};
