// @dynatrace/strato-components/charts/histogram/constants/chart-state.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var chart_state_exports = {};
__export(chart_state_exports, {
  initialState: () => initialState
});
module.exports = __toCommonJS(chart_state_exports);
var import_AnnotationsDataInitialState = require("../../core/components/annotations/providers/AnnotationsDataInitialState.js");
const initialState = {
  filteredSeriesIds: /* @__PURE__ */ new Set(),
  highlightedSeriesId: void 0,
  annotations: import_AnnotationsDataInitialState.initialSelectionState,
  zoom: {
    zoomInEnabled: true,
    zoomOutEnabled: false
  },
  currentDomain: void 0,
  select: void 0,
  interactionMode: "explore",
  explore: {
    visible: false,
    pinned: false,
    position: void 0
  }
};
