// @dynatrace/strato-components/charts/histogram/constants/defaults.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var defaults_exports = {};
__export(defaults_exports, {
  DEFAULT_CHART_HEIGHT: () => DEFAULT_CHART_HEIGHT,
  DEFAULT_LOADING_STATE: () => DEFAULT_LOADING_STATE,
  DEFAULT_SCALE_FALLBACK_COLOR: () => DEFAULT_SCALE_FALLBACK_COLOR,
  DISPLAY_NAME: () => DISPLAY_NAME,
  HISTOGRAM_CHART_ARIA_LABEL: () => HISTOGRAM_CHART_ARIA_LABEL,
  defaultColorPalette: () => defaultColorPalette,
  defaultHistogramChartConfig: () => defaultHistogramChartConfig,
  defaultLegend: () => defaultLegend,
  defaultThreshold: () => defaultThreshold,
  defaultToolbar: () => defaultToolbar,
  defaultToolbarPlacement: () => defaultToolbarPlacement,
  defaultTooltipVariant: () => defaultTooltipVariant,
  defaultXAxis: () => defaultXAxis,
  defaultYAxis: () => defaultYAxis
});
module.exports = __toCommonJS(defaults_exports);
var import_constants = require("../../core/constants.js");
const DISPLAY_NAME = "HistogramChart";
const DEFAULT_CHART_HEIGHT = 300;
const DEFAULT_LOADING_STATE = false;
const defaultLegend = {
  hidden: false,
  position: "auto",
  ratio: import_constants.defaultRatio
};
const defaultColorPalette = "categorical";
const DEFAULT_SCALE_FALLBACK_COLOR = import_constants.SCALE_FALLBACK_COLOR;
const defaultTooltipVariant = "single";
const defaultToolbarPlacement = "top-right";
const defaultToolbar = {
  placement: defaultToolbarPlacement,
  enabled: true
};
const defaultInteractions = {
  zoom: { enabled: true },
  pan: { enabled: true }
};
const defaultHistogramChartConfig = {
  valueRepresentation: import_constants.DEFAULT_VALUE_REPRESENTATION,
  legend: defaultLegend,
  colorPalette: defaultColorPalette,
  tooltip: {
    variant: defaultTooltipVariant
  },
  truncationMode: import_constants.DEFAULT_TRUNCATION_MODE,
  toolbar: defaultToolbar,
  interactions: defaultInteractions
};
const defaultYAxis = {};
const defaultXAxis = {
  scale: "linear"
};
const defaultThreshold = {
  leftThreshold: [],
  rightThreshold: []
};
const HISTOGRAM_CHART_ARIA_LABEL = "Histogram Chart";
