// @dynatrace/strato-components/esm/charts/honeycomb/components/tooltip/HoneycombOverlayTooltip.js
import { jsx } from "react/jsx-runtime";
import { useTooltipData } from "../../../core/components/overlay-tooltip/_to-be-removed/utils.js";
import { OverlayTooltip } from "../../../core/components/overlay-tooltip/OverlayTooltip.js";
import { isTooltipDataEmpty } from "../../../core/components/overlay-tooltip/utils.js";
import { useHoneycombCustomTooltip } from "../../hooks/use-honeycomb-custom-tooltip.js";
import { useHoneycombTruncationMode } from "../../hooks/use-honeycomb-truncation-mode.js";
const HoneycombOverlayTooltip = (props) => {
  const data = useTooltipData().data;
  const truncationMode = useHoneycombTruncationMode();
  const tooltip = useHoneycombCustomTooltip();
  if (tooltip?.hidden || isTooltipDataEmpty(data)) {
    return null;
  }
  const series = data.series[0];
  if (typeof tooltip?.template === "function") {
    const metadata = data.metadata;
    if (metadata) {
      const payload = {
        name: metadata.tooltipMetadata.name,
        value: metadata.tileData.value,
        formattedValue: String(series.value),
        color: metadata.tooltipMetadata.color.Default,
        data: metadata.tileData
      };
      return /* @__PURE__ */ jsx(OverlayTooltip, { legacyTemplate: tooltip.template(payload) });
    }
  }
  return /* @__PURE__ */ jsx(OverlayTooltip, { truncation: truncationMode, children: /* @__PURE__ */ jsx(
    OverlayTooltip.SeriesItem,
    {
      name: series.name,
      value: series.value,
      color: series.color,
      actions: props.actionsForItem?.(series)
    }
  ) });
};
HoneycombOverlayTooltip["displayName"] = "HoneycombOverlayTooltip";
export {
  HoneycombOverlayTooltip
};
