// @dynatrace/strato-components/charts/xy-chart/components/overlay/hooks/useOverlayEvents.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useOverlayEvents_exports = {};
__export(useOverlayEvents_exports, {
  getHoveredDatapoint: () => getHoveredDatapoint,
  useOverlayEvents: () => useOverlayEvents
});
module.exports = __toCommonJS(useOverlayEvents_exports);
var import_useApplyPanDelta = require("./useApplyPanDelta.js");
var import_useApplyZoomDelta = require("./useApplyZoomDelta.js");
var import_useResetInteractions = require("./useResetInteractions.js");
var import_useZoomToSelection = require("./useZoomToSelection.js");
var import_useGridPlotSize = require("../../../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_types = require("../../../../core/components/interactions-overlay/types.js");
var import_useTooltipActions = require("../../../../core/components/overlay-tooltip/useTooltipActions.js");
var import_useSynchronizedCrosshair = require("../../../../core/components/shared-chart-interactions/hooks/useSynchronizedCrosshair.js");
var import_useSynchronizedSelection = require("../../../../core/components/shared-chart-interactions/hooks/useSynchronizedSelection.js");
var import_zoom = require("../../../../core/types/zoom.js");
var import_get_percentage_from_pixel = require("../../../../core/utils/get-percentage-from-pixel.js");
var import_useAreaData = require("../../../hooks/useAreaData.js");
var import_useBarData = require("../../../hooks/useBarData.js");
var import_useDotData = require("../../../hooks/useDotData.js");
var import_useLineData = require("../../../hooks/useLineData.js");
var import_useOverlayPerformance = require("../../../hooks/useOverlayPerformance.js");
var import_useRectData = require("../../../hooks/useRectData.js");
var import_useSyncedXScaleByPriority = require("../../../hooks/useSyncedXScaleByPriority.js");
var import_useXYChartData = require("../../../hooks/useXYChartData.js");
var import_useXYChartId = require("../../../hooks/useXYChartId.js");
var import_reduce_interaction_mode = require("../../../store/reduce-interaction-mode.js");
var import_reducer_explorer = require("../../../store/reducer-explorer.js");
var import_selectors = require("../../../store/selectors.js");
var import_get_prevailing_shape_type = require("../utils/get-prevailing-shape-type.js");
const getHoveredDatapoint = (position, dotSeriesData, lineSeriesData, areaSeriesData, rectData, barData, queryClosest, queryHovered) => {
  const results = {
    dot: queryClosest(dotSeriesData, position.relative),
    line: queryClosest(lineSeriesData, position.relative),
    area: queryClosest(areaSeriesData, position.relative),
    rect: queryHovered(rectData, position.relative),
    bar: queryHovered(barData, position.relative)
  };
  const prevailingShape = (0, import_get_prevailing_shape_type.getPrevailingShapeType)(
    results.dot?.distance,
    results.line?.distance,
    results.area?.distance,
    !!results.rect,
    !!results.bar
  );
  if (!prevailingShape) {
    return void 0;
  }
  return results[prevailingShape];
};
const useOverlayEvents = () => {
  const { filteredSeriesWithDatapoints: rectSeriesData } = (0, import_useRectData.useRectData)();
  const { filteredSeriesWithDatapoints: barSeriesData } = (0, import_useBarData.useBarData)();
  const { filteredSeriesWithDatapoints: dotSeriesData } = (0, import_useDotData.useDotData)();
  const { filteredSeriesWithDatapoints: lineSeriesData } = (0, import_useLineData.useLineData)();
  const { filteredSeriesWithDatapoints: areaSeriesData } = (0, import_useAreaData.useAreaData)();
  const zoomToSelection = (0, import_useZoomToSelection.useZoomToSelection)();
  const setState = (0, import_selectors.useSetState)();
  const mode = (0, import_selectors.useInteractionModeState)();
  const applyPanDelta = (0, import_useApplyPanDelta.useApplyPanDelta)();
  const applyReset = (0, import_useResetInteractions.useResetInteractions)();
  const applyZoomDelta = (0, import_useApplyZoomDelta.useApplyZoomDelta)();
  const zoomState = (0, import_selectors.useZoomState)();
  const { queryClosest, queryHovered } = (0, import_useOverlayPerformance.useOverlayPerformance)();
  const { axesMetadata } = (0, import_useXYChartData.useXYChartData)();
  const chartId = (0, import_useXYChartId.useXYChartId)();
  const synchronizedCrosshair = (0, import_useSynchronizedCrosshair.useSynchronizedCrosshair)();
  const { width } = (0, import_useGridPlotSize.useGridPlotSize)();
  const {
    updateSelectionState: updateSynchronizedSelection,
    resetSelectionState: resetSynchronizedSelection
  } = (0, import_useSynchronizedSelection.useSynchronizedSelection)();
  const syncedXScale = (0, import_useSyncedXScaleByPriority.useSyncedXScaleByPriority)();
  const { hover, pin, unpin, reset, outOfBounds } = (0, import_useTooltipActions.useTooltipActions)();
  const {
    updateHoverState: updateSynchronizedCrosshair,
    setDisabled: disableSynchronizedCrosshair
  } = synchronizedCrosshair ?? {};
  const resetTooltip = () => {
    setState(import_reducer_explorer.reduceExploreCancelHighlight);
    reset();
  };
  const handleModeChange = (event) => {
    const canPan = axesMetadata.isPanEnabled && zoomState.zoomOutEnabled;
    const shouldSwitchToPanMode = event.mode === "pan" && canPan;
    const shouldSwitchToZoomMode = event.mode === "zoom" && axesMetadata.isZoomEnabled;
    if (shouldSwitchToPanMode || shouldSwitchToZoomMode) {
      resetSynchronizedSelection?.();
      setState((0, import_reduce_interaction_mode.reduceInteractionMode)(event.mode));
      setState(import_reducer_explorer.reduceExploreReset);
    } else if (event.mode === "explore") {
      setState((0, import_reduce_interaction_mode.reduceInteractionMode)(import_types.INTERACTION_MODE.Explore));
    }
  };
  const handleHover = ({
    position
  }) => {
    if (syncedXScale) {
      const hoveredDate = syncedXScale.invert(position.relative.x);
      disableSynchronizedCrosshair?.(false);
      updateSynchronizedCrosshair?.(chartId, hoveredDate.valueOf());
    }
    const hoveredDatapointQueryResult = getHoveredDatapoint(
      position,
      dotSeriesData,
      lineSeriesData,
      areaSeriesData,
      rectSeriesData,
      barSeriesData,
      queryClosest,
      queryHovered
    );
    const payload = hoveredDatapointQueryResult ? {
      geometry: hoveredDatapointQueryResult.type,
      header1: void 0,
      series: [],
      metadata: hoveredDatapointQueryResult
    } : void 0;
    hover(
      { clientX: position.absolute.x, clientY: position.absolute.y },
      payload
    );
  };
  const handleTogglePin = (shouldPin, position) => {
    if (shouldPin) {
      handlePin(position);
      updateSynchronizedCrosshair?.(void 0);
    } else {
      handleUnPin(position);
    }
  };
  const handleUnPin = (position) => {
    setState(import_reducer_explorer.reduceExploreUnPin);
    if (position) {
      const hoveredDatapointQueryResult = getHoveredDatapoint(
        position,
        dotSeriesData,
        lineSeriesData,
        areaSeriesData,
        rectSeriesData,
        barSeriesData,
        queryClosest,
        queryHovered
      );
      unpin();
      if (hoveredDatapointQueryResult) {
        hover(
          { clientX: position.absolute.x, clientY: position.absolute.y },
          {
            geometry: hoveredDatapointQueryResult.type,
            header1: void 0,
            series: [],
            metadata: hoveredDatapointQueryResult
          }
        );
      }
    } else {
      resetTooltip();
    }
  };
  const handlePin = (position) => {
    if (!position) {
      return;
    }
    const hoveredDatapointQueryResult = getHoveredDatapoint(
      position,
      dotSeriesData,
      lineSeriesData,
      areaSeriesData,
      rectSeriesData,
      barSeriesData,
      queryClosest,
      queryHovered
    );
    if (hoveredDatapointQueryResult) {
      setState((0, import_reducer_explorer.reduceExplorePin)(hoveredDatapointQueryResult.series.id));
      pin(
        { clientX: position.absolute.x, clientY: position.absolute.y },
        {
          geometry: hoveredDatapointQueryResult.type,
          header1: void 0,
          series: [],
          metadata: hoveredDatapointQueryResult
        }
      );
    }
  };
  const handleReset = (event) => {
    const { mousePosition: position } = event;
    applyReset({ position });
    disableSynchronizedCrosshair?.(false);
    resetSynchronizedSelection?.();
    if (!position) {
      setState(import_reducer_explorer.reduceExploreCancelAll);
      return;
    }
    setState(import_reducer_explorer.reduceExploreCancelAll);
    const hoveredDatapointQueryResult = getHoveredDatapoint(
      position,
      dotSeriesData,
      lineSeriesData,
      areaSeriesData,
      rectSeriesData,
      barSeriesData,
      queryClosest,
      queryHovered
    );
    if (hoveredDatapointQueryResult) {
      hover(
        { clientX: position.absolute.x, clientY: position.absolute.y },
        {
          geometry: hoveredDatapointQueryResult.type,
          header1: void 0,
          series: [],
          metadata: hoveredDatapointQueryResult
        }
      );
    } else {
      outOfBounds();
    }
  };
  const handleLeave = (_) => {
    updateSynchronizedCrosshair?.(void 0);
    outOfBounds();
  };
  const handlePanChange = (event) => {
    applyPanDelta(event.delta);
  };
  const handleSelectStart = (event) => {
    if (mode === "explore") {
      disableSynchronizedCrosshair?.(true);
      updateSynchronizedSelection?.();
    }
    resetTooltip();
  };
  const handleSelectChange = (event) => {
    const variant = event.mode === "explore" ? "select" : "zoom";
    setState(
      (0, import_reducer_explorer.reduceExploreSelect)({
        start: event.start,
        end: event.end,
        finished: false,
        accessible: false,
        variant,
        selectionPercentage: (0, import_get_percentage_from_pixel.getPercentageFromPixel)(
          [event.start, event.end],
          width
        )
      })
    );
    if (mode === "explore" && syncedXScale) {
      updateSynchronizedSelection?.({
        selection: [event.start, event.end],
        chartId,
        scale: syncedXScale
      });
    }
  };
  const handleSelectEnd = (event, ref) => {
    if ((0, import_types.isSelectCancelled)(event)) {
      setState(import_reducer_explorer.reduceExploreCancelSelect);
      resetSynchronizedSelection?.();
    } else if (mode === "explore") {
      setState(
        (0, import_reducer_explorer.reduceExploreSelect)({
          start: event.start,
          end: event.end,
          finished: true,
          accessible: false,
          variant: "select",
          selectionPercentage: (0, import_get_percentage_from_pixel.getPercentageFromPixel)(
            [event.start, event.end],
            width
          )
        })
      );
      ref.current?.focus();
      if (syncedXScale) {
        updateSynchronizedSelection?.({
          selection: [event.start, event.end],
          chartId,
          scale: syncedXScale
        });
      }
    } else {
      zoomToSelection();
    }
  };
  const handleZoomChange = (event) => {
    const canZoomIn = event.behavior === import_zoom.ZOOM_CHANGE_TYPE.ZoomIn && zoomState?.zoomInEnabled;
    const canZoomOut = event.behavior === import_zoom.ZOOM_CHANGE_TYPE.ZoomOut && zoomState.zoomOutEnabled;
    const canZoomX = event.behavior === import_zoom.ZOOM_CHANGE_TYPE.ZoomX;
    const canZoom = canZoomIn || canZoomOut || canZoomX;
    if (canZoom && axesMetadata.isZoomEnabled) {
      setState(import_reducer_explorer.reduceExploreCancelAll);
      applyZoomDelta(event.deltaStart, event.deltaEnd, event.behavior);
    }
  };
  return {
    handleModeChange,
    handleHover,
    handleTogglePin,
    handleReset,
    handleLeave,
    handlePanChange,
    handleSelectChange,
    handleSelectEnd,
    handleZoomChange,
    handleSelectStart,
    resetTooltip
  };
};
