// @dynatrace/strato-components/charts/core/components/tooltip/section-components/VirtualizedTooltipBody/VirtualizedTooltipBody.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var VirtualizedTooltipBody_exports = {};
__export(VirtualizedTooltipBody_exports, {
  VirtualizedTooltipBody: () => VirtualizedTooltipBody
});
module.exports = __toCommonJS(VirtualizedTooltipBody_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_virtual = require("@tanstack/react-virtual");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_VirtualizedTooltipBody_css = require("./VirtualizedTooltipBody.sty.js");
var import_useBodyHeightStyle = require("../../ChartTooltip/hooks/useBodyHeightStyle.js");
var import_constants = require("../../constants.js");
var import_TooltipBody_css = require("../../shared/styles/TooltipBody.sty.js");
const VirtualizedTooltipBody = (props) => {
  const {
    children,
    estimatedItemSize,
    style: consumerStyle,
    className: consumerClassname,
    onScrollContent
  } = props;
  const bodyItems = import_react.Children.toArray(children);
  const bodyItemsCount = bodyItems.length;
  const scrollRef = (0, import_react.useRef)(null);
  const { getVirtualItems, getTotalSize, measureElement } = (0, import_react_virtual.useVirtualizer)({
    count: bodyItemsCount,
    getScrollElement: () => scrollRef.current,
    estimateSize: (index) => {
      return estimatedItemSize?.(index) ?? import_constants.ESTIMATED_ITEM_SIZE;
    }
  });
  const items = getVirtualItems();
  const totalSize = getTotalSize();
  const bodyHeightStyle = (0, import_useBodyHeightStyle.useBodyHeightStyle)(totalSize, bodyItemsCount);
  (0, import_react.useEffect)(() => {
    if (!onScrollContent) {
      return;
    }
    const scrollContainer = scrollRef.current;
    scrollContainer?.addEventListener("scroll", onScrollContent);
    return () => {
      scrollContainer?.removeEventListener("scroll", onScrollContent);
    };
  }, [onScrollContent]);
  if (!bodyItemsCount) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      className: (0, import_clsx.default)(import_TooltipBody_css.container, consumerClassname),
      ref: scrollRef,
      style: { ...bodyHeightStyle, ...consumerStyle },
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { style: { height: totalSize }, className: import_VirtualizedTooltipBody_css.listContainer, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        "ul",
        {
          className: import_VirtualizedTooltipBody_css.itemsList,
          style: { transform: `translateY(${items[0]?.start ?? 0}px)` },
          "data-testid": "virtualized-items-list",
          children: items.map((item) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)("li", { "data-index": item.index, ref: measureElement, children: bodyItems[item.index] }, item.key))
        }
      ) })
    }
  );
};
