// @dynatrace/strato-components/esm/tables/DataTable/features/Thresholds/default-comparators.js
function greaterThanComparator(ruleValue, compareValue) {
  if (typeof ruleValue === "number" && typeof compareValue === "number") {
    return compareValue > ruleValue;
  }
  return false;
}
function lessThanComparator(ruleValue, compareValue) {
  if (typeof ruleValue === "number" && typeof compareValue === "number") {
    return compareValue < ruleValue;
  }
  return false;
}
function greaterThanOrEqualToComparator(ruleValue, compareValue) {
  if (typeof ruleValue === "number" && typeof compareValue === "number") {
    return compareValue >= ruleValue;
  }
  return false;
}
function lessThanOrEqualToComparator(ruleValue, compareValue) {
  if (typeof ruleValue === "number" && typeof compareValue === "number") {
    return compareValue <= ruleValue;
  }
  return false;
}
function equalToComparator(ruleValue, compareValue) {
  if (typeof ruleValue === "number" && typeof compareValue === "number") {
    return ruleValue === compareValue;
  }
  if (typeof ruleValue === "string" && typeof compareValue === "string") {
    return ruleValue.localeCompare(compareValue) === 0;
  }
  return false;
}
function notEqualToComparator(ruleValue, compareValue) {
  if (typeof ruleValue === "number" && typeof compareValue === "number") {
    return ruleValue !== compareValue;
  }
  if (typeof ruleValue === "string" && typeof compareValue === "string") {
    return ruleValue.localeCompare(compareValue) !== 0;
  }
  return ruleValue !== compareValue;
}
const defaultComparators = {
  "greater-than": greaterThanComparator,
  "less-than": lessThanComparator,
  "greater-than-or-equal-to": greaterThanOrEqualToComparator,
  "less-than-or-equal-to": lessThanOrEqualToComparator,
  "equal-to": equalToComparator,
  "not-equal-to": notEqualToComparator
};
export {
  defaultComparators
};
