// @dynatrace/strato-components/esm/tables/DataTable/features/Thresholds/normalize-rules.js
import { makeAccessorFunction } from "../../hooks/makeAccessorFunction.js";
function makeAccessor(accessor, fallbackAccessor) {
  const appliedAccessor = accessor ?? fallbackAccessor;
  if (typeof appliedAccessor === "string") {
    if (appliedAccessor.includes(".") || appliedAccessor.includes("[")) {
      return makeAccessorFunction(appliedAccessor);
    } else {
      return (row) => {
        return row[appliedAccessor];
      };
    }
  } else if (typeof appliedAccessor === "function") {
    return appliedAccessor;
  }
  return () => void 0;
}
function normalizeThresholdRules(thresholds, fallbackAccessor) {
  const columnThresholdDefinitions = [];
  for (const threshold of thresholds) {
    const unifiedRule = {
      color: threshold.color,
      backgroundColor: "backgroundColor" in threshold ? threshold.backgroundColor : void 0,
      rules: [],
      type: "type" in threshold ? threshold.type : void 0
    };
    const normalizedRulesArray = threshold.rules !== void 0 ? threshold.rules : [
      {
        accessor: "accessor" in threshold ? threshold.accessor : void 0,
        value: "value" in threshold ? threshold.value : void 0,
        comparator: threshold.comparator
      }
    ];
    unifiedRule.rules = normalizedRulesArray.map((rule) => {
      return {
        // Accessors might be undefined in scenarios where a custom comparator function
        // is given.
        accessor: "accessor" in rule ? makeAccessor(rule.accessor, fallbackAccessor) : void 0,
        // Value might be undefined in scenarios where a custom comparator function
        // is given.
        value: "value" in rule ? rule.value : void 0,
        comparator: rule.comparator
      };
    });
    columnThresholdDefinitions.push(unifiedRule);
  }
  return columnThresholdDefinitions;
}
export {
  normalizeThresholdRules
};
