// @dynatrace/strato-components/esm/tables/DataTable/features/Thresholds/Thresholds.js
import { defaultComparators } from "./default-comparators.js";
import { normalizeThresholdRules } from "./normalize-rules.js";
import { useDataTableConfigElement } from "../../../hooks/useDataTableConfig.js";
import {
  isAccessorFnColumnDef,
  isAccessorKeyColumnDef,
  isGroupColumnDef
} from "../../hooks/useTable/datatable-typeguards.js";
import {
  assignRowMethods,
  getMemoOptions
} from "../../hooks/useTable/utils.js";
import { memo } from "../../tanstack-table-core-override/src/index.js";
function useThresholds(props, options) {
  const configThresholds = useDataTableConfigElement("rowThresholds");
  const rowThresholds = props.rowThresholds ?? configThresholds;
  if (rowThresholds) {
    options.rowThresholds = normalizeThresholdRules(rowThresholds);
    options.hasThresholds = true;
  } else {
    options.rowThresholds = void 0;
    options.hasThresholds = options.columns.some((column) => {
      if (isGroupColumnDef(column)) {
        const groupedColumn = column;
        return groupedColumn.columns && groupedColumn.columns.some(
          (childColumn) => childColumn.thresholds && childColumn.thresholds.length > 0
        );
      }
      return column.thresholds && column.thresholds.length > 0;
    });
  }
}
function matchesThreshold(threshold, rowData) {
  return threshold.rules.every((rule) => {
    const compareValue = rule.accessor?.(rowData) ?? void 0;
    if (typeof rule.comparator === "string") {
      return defaultComparators[rule.comparator](rule.value, compareValue);
    }
    return rule.comparator(rowData);
  });
}
const Thresholds = {
  createTable(table) {
    assignRowMethods(table, {
      getThresholdPill: memo(
        function() {
          return [table.options.rowThresholds, this.id, this.index];
        },
        function(rowThresholds) {
          if (rowThresholds === void 0 || rowThresholds.length === 0) {
            return null;
          }
          const pillColor = rowThresholds.findLast(
            (threshold) => threshold.type === "pill" && matchesThreshold(threshold, this.original)
          )?.color;
          return pillColor ?? null;
        },
        getMemoOptions(table.options, "debugRows", "row.getThresholdPill")
      ),
      getThresholdColors: memo(
        function() {
          return [table.options.rowThresholds, this.id, this.index];
        },
        function(rowThresholds) {
          if (rowThresholds === void 0 || rowThresholds.length === 0) {
            return null;
          }
          const matchedThreshold = rowThresholds.findLast(
            (threshold) => threshold.type === "highlight" && matchesThreshold(threshold, this.original)
          );
          if (matchedThreshold) {
            return {
              color: matchedThreshold.color,
              backgroundColor: matchedThreshold.backgroundColor
            };
          }
          return null;
        },
        getMemoOptions(table.options, "debugRows", "row.getThresholdColors")
      )
    });
  },
  createColumn(column, table) {
    column.getThresholdRules = memo(
      () => [column.columnDef.thresholds],
      (thresholds) => {
        if (thresholds === void 0 || thresholds.length === 0) {
          return null;
        }
        const columnDef = column.columnDef;
        let fallbackAccessor;
        if (isAccessorKeyColumnDef(columnDef)) {
          fallbackAccessor = columnDef.accessorKey;
        }
        if (isAccessorFnColumnDef(columnDef)) {
          const accessorFn = columnDef.accessorFn;
          fallbackAccessor = (row) => accessorFn(row, 0);
        }
        return normalizeThresholdRules(thresholds, fallbackAccessor);
      },
      getMemoOptions(table.options, "debugColumns", "column.getThresholdRules")
    );
  },
  createCell(cell, column, row) {
    cell.getThresholdColors = () => {
      const columnThresholds = column.getThresholdRules();
      const matchedColors = {
        color: void 0,
        backgroundColor: void 0
      };
      if (columnThresholds) {
        const matchedColumnThreshold = columnThresholds.findLast(
          (threshold) => matchesThreshold(threshold, row.original)
        );
        if (matchedColumnThreshold) {
          matchedColors.color = matchedColumnThreshold.color;
          matchedColors.backgroundColor = matchedColumnThreshold.backgroundColor;
          return matchedColors;
        }
      }
      const rowThresholds = row.getThresholdColors();
      if (rowThresholds) {
        matchedColors.color = rowThresholds.color;
        matchedColors.backgroundColor = rowThresholds.backgroundColor;
      }
      return matchedColors;
    };
  }
};
export {
  Thresholds,
  useThresholds
};
