// @dynatrace/strato-components/esm/typography/text-ellipsis/TextEllipsis.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import {
  forwardRef,
  useLayoutEffect,
  useRef
} from "react";
import useResizeObserver from "use-resize-observer";
import { useAriaLabelingProps } from "../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useFontsUpdated } from "../../core/hooks/useFontsUpdated.js";
import { useMergeRefs } from "../../core/hooks/useMergeRefs.js";
import { useSafeForwardProps } from "../../core/hooks/useSafeForwardProps.js";
import { centerEllipsizeText } from "../utils.js";
import * as styles from "./TextEllipsis.sty.js";
const COMPONENT_NAME = "TextEllipsis";
function getFontProps(innerSpanElement) {
  const computedStyle = window.getComputedStyle(innerSpanElement, null);
  return {
    fontStyle: computedStyle.getPropertyValue("font-style"),
    fontWeight: computedStyle.getPropertyValue("font-weight"),
    fontSize: computedStyle.getPropertyValue("font-size"),
    fontFamily: computedStyle.getPropertyValue("font-family")
  };
}
function isNativeEllipsisActive(innerSpanElement, containerWidth) {
  const originalTextWidth = Math.max(
    innerSpanElement.getBoundingClientRect().width,
    innerSpanElement.scrollWidth
  );
  const textWidth = Math.floor(originalTextWidth * 100) / 100;
  return textWidth > containerWidth;
}
function resetFullText(innerSpanElement, originalText) {
  if (innerSpanElement.textContent !== originalText) {
    innerSpanElement.textContent = originalText;
  }
}
function getWidth(element) {
  return Math.max(element.getBoundingClientRect().width, element.clientWidth);
}
function getContainerWidth(containerElement) {
  return Math.ceil(getWidth(containerElement) * 100) / 100;
}
function onTextOverflowCallback(current, wasEllipsized, onTextOverflow) {
  if (onTextOverflow && current !== wasEllipsized) {
    onTextOverflow(current);
  }
}
function renderEllipsizedText(originalText, truncationMode, containerRef, wasEllipsized, onTextOverflow) {
  const containerElement = containerRef.current;
  const innerSpanElement = containerElement?.firstElementChild || null;
  if (containerElement === null || innerSpanElement === null) {
    return false;
  }
  resetFullText(innerSpanElement, originalText);
  const containerWidth = getContainerWidth(containerElement);
  if (!isNativeEllipsisActive(innerSpanElement, containerWidth)) {
    onTextOverflowCallback(false, wasEllipsized, onTextOverflow);
    return false;
  }
  if (truncationMode === "middle") {
    innerSpanElement.textContent = centerEllipsizeText(
      originalText,
      getFontProps(innerSpanElement),
      containerWidth
    );
  }
  onTextOverflowCallback(true, wasEllipsized, onTextOverflow);
  return true;
}
const ComputedTextEllipsis = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const {
    children,
    truncationMode = "end",
    onTextOverflow,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
  const ariaLabelingProps = useAriaLabelingProps(restProps);
  const forwardProps = useSafeForwardProps(restProps, ariaLabelingProps);
  const { ref: observerContainerRef, width } = useResizeObserver();
  const containerRef = useRef(null);
  const ellipsisStateRef = useRef(false);
  const mergedRef = useMergeRefs([
    observerContainerRef,
    (instance) => {
      containerRef.current = instance;
    },
    forwardedRef
  ]);
  const fontsUpdated = useFontsUpdated();
  useLayoutEffect(() => {
    ellipsisStateRef.current = renderEllipsizedText(
      children,
      truncationMode,
      containerRef,
      ellipsisStateRef.current,
      onTextOverflow
    );
  }, [width, children, truncationMode, fontsUpdated, onTextOverflow]);
  return /* @__PURE__ */ jsx(
    "span",
    {
      ref: mergedRef,
      className: clsx(
        styles.ellipsis({ truncationMode }),
        {
          [styles.active]: ellipsisStateRef.current
        },
        consumerClassName
      ),
      style: consumerStyle,
      "data-sourcetext": children,
      "data-ellipsis": true,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...ariaLabelingProps,
      ...behavioralTrackingProps,
      ...forwardProps,
      "data-dt-component": COMPONENT_NAME,
      children: /* @__PURE__ */ jsx("span", { children })
    }
  );
});
const CssTextEllipsis = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const {
    children,
    truncationMode = "end",
    onTextOverflow,
    className: consumerClassName,
    style: consumerStyle,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "data-testid": dataTestId,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
  const ariaLabelingProps = useAriaLabelingProps(restProps);
  const forwardProps = useSafeForwardProps(restProps, ariaLabelingProps);
  return /* @__PURE__ */ jsx(
    "span",
    {
      ref: forwardedRef,
      "data-ellipsis": true,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      className: clsx(styles.ellipsis({ truncationMode }), consumerClassName),
      style: consumerStyle,
      ...ariaLabelingProps,
      ...behavioralTrackingProps,
      ...forwardProps,
      "data-dt-component": COMPONENT_NAME,
      children: /* @__PURE__ */ jsx("span", { children })
    }
  );
});
const TextEllipsis = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const cssOnlyEllipsis = !props.onTextOverflow && props.truncationMode !== "middle";
  return cssOnlyEllipsis ? /* @__PURE__ */ jsx(CssTextEllipsis, { ref: forwardedRef, ...props }) : /* @__PURE__ */ jsx(ComputedTextEllipsis, { ref: forwardedRef, ...props });
});
TextEllipsis.displayName = COMPONENT_NAME;
export {
  TextEllipsis
};
