// @dynatrace/strato-components/esm/charts/core/components/chart-layout/components/ResizableChartLayout.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import {
  graphContainer,
  legendContainer,
  legendResizeLine,
  resizeContainer
} from "./styles.sty.js";
import { Flex } from "../../../../../layouts/flex/Flex.js";
import { LegendSizeContext } from "../../legend/shared/context/LegendSizeContext.js";
import { GraphSizeContext } from "../context/GraphSize.context.js";
import { useChartLayoutLoading } from "../hooks/useChartLayoutLoading.js";
import { useChartLayoutOrientation } from "../hooks/useChartLayoutOrientation.js";
import { useChartLayoutSize } from "../hooks/useChartLayoutSize.js";
import { useChartLayoutTemplate } from "../hooks/useChartLayoutTemplate.js";
import { useResizableLayout } from "../hooks/useResizableLayout.js";
import { isVertical as isVerticalFn } from "../utils/orientation.js";
const ResizableChartLayout = ({
  onResize,
  showResizer,
  showLegend
}) => {
  const { orientation } = useChartLayoutOrientation();
  const loading = useChartLayoutLoading();
  const containerUsableArea = useChartLayoutSize();
  const { legendArea, graphArea } = useChartLayoutTemplate();
  const isVertical = isVerticalFn(orientation);
  const {
    legendWidth,
    legendHeight,
    graphSize,
    legendContainerRef,
    legendResizeLineRef,
    legendResizeContainerRef
  } = useResizableLayout(
    containerUsableArea.width,
    containerUsableArea.height,
    orientation,
    showResizer,
    showLegend,
    onResize
  );
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx("div", { className: graphContainer({ loading }), children: /* @__PURE__ */ jsx(GraphSizeContext.Provider, { value: graphSize, children: graphArea }) }),
    showLegend && showResizer && /* @__PURE__ */ jsx(
      Flex,
      {
        ref: legendResizeContainerRef,
        alignItems: "center",
        justifyContent: isVertical ? "center" : "flex-end",
        className: resizeContainer({ orientation, loading }),
        children: /* @__PURE__ */ jsx(
          "div",
          {
            "data-testid": "legend-resize-line",
            ref: legendResizeLineRef,
            className: legendResizeLine({ orientation })
          }
        )
      }
    ),
    showLegend && /* @__PURE__ */ jsx(
      "div",
      {
        ref: legendContainerRef,
        className: legendContainer({
          orientation,
          loading
        }),
        children: /* @__PURE__ */ jsx(
          LegendSizeContext.Provider,
          {
            value: { width: legendWidth, height: legendHeight },
            children: legendArea
          }
        )
      }
    )
  ] });
};
ResizableChartLayout["displayName"] = "ResizableChartLayout";
export {
  ResizableChartLayout
};
