// @dynatrace/strato-components/esm/notifications/notification-settings/components/Content.js
import { jsx, jsxs } from "react/jsx-runtime";
import { forwardRef, useContext, useMemo } from "react";
import { isElement } from "react-is";
import { NotifyButtonContext } from "../../../buttons/notify-button/contexts/NotifyButtonContext.js";
import { canReceiveDefaultTextFormat } from "../../../core/utils/can-receive-default-text-format.js";
import { Flex } from "../../../layouts/flex/Flex.js";
import { Text } from "../../../typography/text/Text.js";
import { NotificationSettingsContext } from "../contexts/NotificationSettingsContext.js";
import { useNotificationMessages } from "../hooks/useNotificationMessages.js";
import * as styles from "../NotificationSettings.sty.js";
const Content = forwardRef((props, ref) => {
  const { children, ...remainingProps } = props;
  const { overlayAriaLabelledBy } = useContext(NotifyButtonContext);
  const { enabled, hasReadPermission, hasWritePermission } = useContext(
    NotificationSettingsContext
  );
  const { headingText, infoText } = useNotificationMessages({
    notificationEnabled: enabled
  });
  const childrenContent = useMemo(() => {
    if (typeof children === "function") {
      return children({
        disabled: !enabled,
        headingText,
        infoText,
        hasReadPermission,
        hasWritePermission
      });
    }
    return children;
  }, [
    children,
    enabled,
    hasReadPermission,
    hasWritePermission,
    headingText,
    infoText
  ]);
  if (children === void 0) {
    return /* @__PURE__ */ jsxs(Flex, { ref, flexDirection: "column", gap: 0, ...remainingProps, children: [
      /* @__PURE__ */ jsx(Text, { id: overlayAriaLabelledBy, textStyle: "base-emphasized", children: headingText }),
      /* @__PURE__ */ jsx(Text, { textStyle: "small", className: styles.overlayText, children: infoText })
    ] });
  }
  if (canReceiveDefaultTextFormat(childrenContent)) {
    return /* @__PURE__ */ jsx(
      Text,
      {
        id: overlayAriaLabelledBy,
        ref,
        as: "div",
        textStyle: "base",
        ...remainingProps,
        children: childrenContent
      }
    );
  }
  if (isElement(childrenContent)) {
    return childrenContent;
  }
  return /* @__PURE__ */ jsx("div", { ref, ...remainingProps, children: childrenContent });
});
export {
  Content
};
