// @dynatrace/strato-components/esm/notifications/notification-settings/components/ToggleButton.js
import { jsx } from "react/jsx-runtime";
import { useContext } from "react";
import { FormattedMessage } from "react-intl";
import { Button } from "../../../buttons/button/Button.js";
import { NotificationSettingsContext } from "../contexts/NotificationSettingsContext.js";
import { notificationMessages } from "../messages.js";
function NotificationToggleButton() {
  const {
    enabled,
    loading,
    toggleNotifications,
    hasReadPermission,
    hasNotFoundError
  } = useContext(NotificationSettingsContext);
  return /* @__PURE__ */ jsx(
    Button,
    {
      variant: enabled ? "emphasized" : "accent",
      color: enabled ? "neutral" : "primary",
      onClick: toggleNotifications,
      disabled: !hasReadPermission || hasNotFoundError,
      loading,
      children: /* @__PURE__ */ jsx(
        FormattedMessage,
        {
          ...notificationMessages.notificationsToggle,
          values: { notificationEnabled: enabled }
        }
      )
    }
  );
}
export {
  NotificationToggleButton
};
