// @dynatrace/strato-components/editors/dql-editor/actions-menu/actions-menu-messages.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var actions_menu_messages_exports = {};
__export(actions_menu_messages_exports, {
  dqlEditorActionsMenuMessages: () => dqlEditorActionsMenuMessages
});
module.exports = __toCommonJS(actions_menu_messages_exports);
var import_react_intl = require("react-intl");
const dqlEditorActionsMenuMessages = (0, import_react_intl.defineMessages)({
  triggerLabel: {
    defaultMessage: "DQL editor actions",
    id: "VYHp9Wzie7wOmBmf",
    description: "Aria label for the DQL editor actions menu trigger button."
  },
  copyQuery: {
    defaultMessage: "Copy query",
    id: "NwDE2g2nYFjxPG9v",
    description: "Menu item label to copy the DQL query to the clipboard."
  },
  openDocumentation: {
    defaultMessage: "Open documentation",
    id: "XmYqCKn9iEfBoSSL",
    description: "Menu item label to open the DQL documentation in a new browser tab."
  }
});
