// @dynatrace/strato-components/editors/dql-editor/actions-menu/ActionsMenu.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ActionsMenu_exports = {};
__export(ActionsMenu_exports, {
  ActionsMenu: () => ActionsMenu,
  ActionsMenuInternal: () => ActionsMenuInternal
});
module.exports = __toCommonJS(ActionsMenu_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_actions_menu_messages = require("./actions-menu-messages.js");
var styles = __toESM(require("./ActionsMenu.sty.js"));
var import_Button = require("../../../buttons/button/Button.js");
var import_external_links = require("../../../core/external-links.js");
var import_Menu = require("../../../navigation/menu/Menu.js");
const ActionsMenu = () => null;
ActionsMenu.displayName = "DQLEditor.ActionsMenu";
const ActionsMenuInternal = ({
  getValue,
  size
}) => {
  const intl = (0, import_react_intl.useIntl)();
  const copyToClipboard = (0, import_react.useCallback)(async () => {
    try {
      await navigator.clipboard.writeText(getValue());
    } catch {
    }
  }, [getValue]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: styles.actionsMenuWrapper({ size }), children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Menu.Menu, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Trigger, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Button.Button,
      {
        variant: "default",
        size,
        "aria-label": intl.formatMessage(
          import_actions_menu_messages.dqlEditorActionsMenuMessages.triggerLabel
        ),
        "data-dt-name": "OpenActionsMenu",
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.DotMenuIcon, {}) })
      }
    ) }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Menu.Menu.Content, { children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Group, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
        import_Menu.Menu.Item,
        {
          onSelect: () => {
            copyToClipboard().catch(() => void 0);
          },
          "data-dt-name": "CopyQuery",
          children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.CopyIcon, {}) }),
            intl.formatMessage(import_actions_menu_messages.dqlEditorActionsMenuMessages.copyQuery)
          ]
        }
      ) }),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Group, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
        import_Menu.Menu.Link,
        {
          href: import_external_links.DQL_DOCUMENTATION_URL,
          target: "_blank",
          rel: "noopener noreferrer",
          "data-dt-name": "OpenDocumentation",
          children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ExternalLinkIcon, {}) }),
            intl.formatMessage(
              import_actions_menu_messages.dqlEditorActionsMenuMessages.openDocumentation
            )
          ]
        }
      ) })
    ] })
  ] }) });
};
