// @dynatrace-sdk/client-state/esm/index.js
/**
 * @license
 * Copyright 2023-2025 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// packages/client/state/src/lib/apis/state-api.ts
import {
  httpClient as defaultHttpClient,
  isHttpClientAbortError,
  isHttpClientResponseError
} from "@dynatrace-sdk/http-client";
import {
  ApiClientError,
  InvalidResponseError,
  isInvalidResponseError
} from "@dynatrace-sdk/shared-errors";

// packages/client/state/src/lib/error-envelopes/api-gateway-errors-handler.ts
import { ApiGatewayError } from "@dynatrace-sdk/shared-errors";
async function apiGatewayErrorsHandler(response) {
  if (response.headers?.["dynatrace-response-source"] === "API Gateway") {
    const responseBody = await response.body("json");
    throw new ApiGatewayError(response, responseBody);
  }
}

// packages/client/state/src/lib/error-envelopes/get-error-message.ts
function serializeData(data) {
  try {
    return JSON.stringify(data);
  } catch (e) {
    return String(data);
  }
}
function getMessagesFromErrorDetails(details) {
  const messages = [];
  Object.entries(details).forEach(([name, data]) => {
    if (!data)
      return;
    const serializedData = serializeData(data);
    switch (name) {
      case "missingScopes":
        messages.push(`Missing scopes: ${serializedData}`);
        break;
      default:
        messages.push(`${name}: ${serializedData}`);
    }
  }, []);
  return messages;
}
function getErrorMessage(errorBody, defaultMessage) {
  const error = errorBody;
  const msg = error?.error?.message || defaultMessage;
  const details = error?.error?.details || {};
  return [msg, ...getMessagesFromErrorDetails(details)].join(". ");
}

// packages/client/state/src/lib/error-envelopes/error-response-error.ts
import { ClientRequestError } from "@dynatrace-sdk/shared-errors";
var ErrorResponseError = class extends ClientRequestError {
  isErrorResponseError = true;
};
function isErrorResponseError(e) {
  return e?.isApiClientError === true && e?.isClientRequestError === true && e?.isErrorResponseError === true && e instanceof Error;
}

// packages/client/state/src/lib/error-envelopes/responses/app-state-limits-exceeded.ts
var AppStateLimitsExceeded = class extends ErrorResponseError {
  isAppStateLimitsExceeded = true;
  name = "AppStateLimitsExceeded";
};
function isAppStateLimitsExceeded(e) {
  return e?.isApiClientError === true && e?.isClientRequestError === true && e?.isErrorResponseError === true && e?.isAppStateLimitsExceeded === true && e instanceof Error;
}

// packages/client/state/src/lib/error-envelopes/responses/forbidden.ts
var Forbidden = class extends ErrorResponseError {
  isForbidden = true;
  name = "Forbidden";
};
function isForbidden(e) {
  return e?.isApiClientError === true && e?.isClientRequestError === true && e?.isErrorResponseError === true && e?.isForbidden === true && e instanceof Error;
}

// packages/client/state/src/lib/error-envelopes/responses/general-error.ts
var GeneralError = class extends ErrorResponseError {
  isGeneralError = true;
  name = "GeneralError";
};
function isGeneralError(e) {
  return e?.isApiClientError === true && e?.isClientRequestError === true && e?.isErrorResponseError === true && e?.isGeneralError === true && e instanceof Error;
}

// packages/client/state/src/lib/error-envelopes/responses/internal-server-error.ts
var InternalServerError = class extends ErrorResponseError {
  isInternalServerError = true;
  name = "InternalServerError";
};
function isInternalServerError(e) {
  return e?.isApiClientError === true && e?.isClientRequestError === true && e?.isErrorResponseError === true && e?.isInternalServerError === true && e instanceof Error;
}

// packages/client/state/src/lib/error-envelopes/responses/not-found.ts
var NotFound = class extends ErrorResponseError {
  isNotFound = true;
  name = "NotFound";
};
function isNotFound(e) {
  return e?.isApiClientError === true && e?.isClientRequestError === true && e?.isErrorResponseError === true && e?.isNotFound === true && e instanceof Error;
}

// packages/client/state/src/lib/error-envelopes/responses/unauthorized.ts
var Unauthorized = class extends ErrorResponseError {
  isUnauthorized = true;
  name = "Unauthorized";
};
function isUnauthorized(e) {
  return e?.isApiClientError === true && e?.isClientRequestError === true && e?.isErrorResponseError === true && e?.isUnauthorized === true && e instanceof Error;
}

// packages/client/state/src/lib/error-envelopes/responses/user-app-state-limits-exceeded.ts
var UserAppStateLimitsExceeded = class extends ErrorResponseError {
  isUserAppStateLimitsExceeded = true;
  name = "UserAppStateLimitsExceeded";
};
function isUserAppStateLimitsExceeded(e) {
  return e?.isApiClientError === true && e?.isClientRequestError === true && e?.isErrorResponseError === true && e?.isUserAppStateLimitsExceeded === true && e instanceof Error;
}

// packages/client/state/src/lib/utils/transformations.ts
var defaultRuleConfig = {
  force: false,
  datetime: false
};
function mergeKeys(key1, key2) {
  if (!key1)
    return key2;
  if (!key2)
    return key1;
  return `${key1}.${key2}`;
}
function matchKey(key, rules, cache) {
  if (cache.has(key)) {
    return cache.get(key);
  }
  for (const rule of rules) {
    const pattern = Array.isArray(rule) ? rule[0] : rule;
    const config = Array.isArray(rule) ? rule[1] : defaultRuleConfig;
    if (key === pattern) {
      const result = [pattern, config];
      cache.set(key, result);
      return result;
    }
    let i = 0;
    let j = 0;
    const keyLen = key.length;
    const patternLen = pattern.length;
    let isMatch = true;
    while (i < keyLen && j < patternLen) {
      let keyPartEnd = key.indexOf(".", i);
      let patternPartEnd = pattern.indexOf(".", j);
      if (keyPartEnd === -1)
        keyPartEnd = keyLen;
      if (patternPartEnd === -1)
        patternPartEnd = patternLen;
      const keyPart = key.slice(i, keyPartEnd);
      const patternPart = pattern.slice(j, patternPartEnd);
      if (patternPart !== "*" && patternPart !== "**" && keyPart !== patternPart) {
        isMatch = false;
        break;
      }
      i = keyPartEnd + 1;
      j = patternPartEnd + 1;
    }
    if (isMatch) {
      if (j < patternLen) {
        const remaining = pattern.slice(j);
        if (remaining !== "**" && remaining !== "*") {
          isMatch = false;
        }
      }
      if (i < keyLen && pattern.slice(j - 1) !== "**") {
        isMatch = false;
      }
    }
    if (isMatch) {
      const result = [pattern, config];
      cache.set(key, result);
      return result;
    }
  }
  cache.set(key, null);
  return null;
}
var potentialMatchForRoot = (pattern, key) => pattern.startsWith(key) || pattern.startsWith("*") || pattern.startsWith("**");
var potentialMatch = (pattern, key) => pattern.startsWith(key) || pattern.includes("*") || pattern.includes("**");
function partialMatch(key, keyPatterns) {
  const keyPatternsLen = keyPatterns.length;
  if (!key.includes(".")) {
    for (let i = 0; i < keyPatternsLen; i++) {
      const pattern = keyPatterns[i];
      if (potentialMatchForRoot(pattern, key))
        return true;
    }
    return false;
  }
  let dotIndex = -1;
  let currentKeyEnd = 0;
  while ((dotIndex = key.indexOf(".", currentKeyEnd)) !== -1) {
    const currentKey = key.slice(0, dotIndex);
    const isRoot = !currentKey.includes(".");
    for (let i = 0; i < keyPatternsLen; i++) {
      const pattern = keyPatterns[i];
      if (isRoot) {
        if (potentialMatchForRoot(pattern, currentKey))
          return true;
      } else if (potentialMatch(pattern, currentKey))
        return true;
    }
    currentKeyEnd = dotIndex + 1;
  }
  for (let i = 0; i < keyPatternsLen; i++) {
    const pattern = keyPatterns[i];
    if (potentialMatch(pattern, key))
      return true;
  }
  return false;
}
var dateTimePattern = /^(\d{4}-\d{2}-\d{2})(T\d{2}:\d{2}:\d{2}(?:\.\d+)?(Z|[+-]\d{2}:\d{2})?)?$/i;
function isDateLike(value) {
  if (typeof value === "string") {
    return dateTimePattern.test(value);
  }
  return false;
}
function transformValue(value, rules, direction) {
  const { force, datetime } = rules;
  if (direction === "to") {
    return force || isDateLike(value) ? new Date(value) : value;
  }
  if (value instanceof Date) {
    const iso = value.toISOString();
    return datetime ? iso : iso.split("T")[0];
  }
  return value;
}
function walk(obj, direction, rules, rulesCache, keyPatterns, currentKey = "") {
  if (rules.length <= 0) {
    return obj;
  }
  if (!partialMatch(currentKey, keyPatterns)) {
    return obj;
  }
  if (Array.isArray(obj)) {
    const match = matchKey(currentKey, rules, rulesCache);
    obj.forEach((item, idx) => {
      if (typeof item === "object" && item !== null) {
        walk(item, direction, rules, rulesCache, keyPatterns, currentKey);
      } else if (match) {
        obj[idx] = transformValue(obj[idx], match[1], direction);
      }
    });
  }
  if (typeof obj === "object" && obj !== null && !Array.isArray(obj)) {
    for (const key of Object.keys(obj)) {
      const mergedKey = mergeKeys(currentKey, key);
      const match = matchKey(mergedKey, rules, rulesCache);
      if (match) {
        obj[key] = transformValue(obj[key], match[1], direction);
      }
      walk(obj[key], direction, rules, rulesCache, keyPatterns, mergedKey);
    }
  }
  return obj;
}
function flattenKeyPatterns(keyPatterns) {
  return keyPatterns.map((k) => Array.isArray(k) ? k[0] : k);
}
function transform(direction, object, rules) {
  const flatKeyPatterns = flattenKeyPatterns(rules);
  const rulesCache = /* @__PURE__ */ new Map();
  return walk(object, direction, rules, rulesCache, flatKeyPatterns);
}
function transformRequest(object, rules) {
  return transform("from", object, rules);
}
function transformResponse(object, rules) {
  return transform("to", object, rules);
}

// packages/client/state/src/lib/utils/url-helpers.ts
var encodeQueryParam = (key, value) => {
  const encodedKey = encodeURIComponent(key);
  return `${encodedKey}=${encodeURIComponent(typeof value === "number" ? value : String(value))}`;
};
var addExplodedArrayQueryParam = (query, key) => {
  const arrayValue = query[key];
  return arrayValue.map((value) => encodeQueryParam(key, value)).join("&");
};
var addNonExplodedArrayQueryParams = (query, key) => {
  const encodedKey = encodeURIComponent(key);
  const encodedParamsList = query[key].map((value) => encodeURIComponent(typeof value === "number" ? value : String(value))).join(",");
  return `${encodedKey}=${encodedParamsList}`;
};
var addQueryParam = (query, key) => encodeQueryParam(key, query[key]);
var arrayQueryParams = (query, key, explode) => {
  return explode ? addExplodedArrayQueryParam(query, key) : addNonExplodedArrayQueryParams(query, key);
};
var toQueryString = (rawQuery, flags = {}) => {
  const query = rawQuery || {};
  const keys = Object.keys(query).filter((key) => typeof query[key] !== "undefined");
  const queryString = keys.map(
    (key) => Array.isArray(query[key]) ? arrayQueryParams(query, key, flags.explode?.hasOwnProperty(key) ? flags.explode[key] : true) : addQueryParam(query, key)
  ).join("&");
  return queryString ? `?${queryString}` : "";
};

// packages/client/state/src/lib/apis/state-api.ts
var StateClient = class {
  httpClient;
  shouldTransformDates = true;
  /**
   * @param {HttpClient} httpClientImplementation - You can provide custom http client as a parameter to constructor of a StateClient. Custom http client should implement {HttpClient} interface.
   * @param {ApiClientOptions} [options] - Optional configuration for SDK Client.
   * @example
   * const stateClientNodeJs = new StateClient(yourCustomImplementation);
   */
  constructor(httpClientImplementation, options) {
    this.httpClient = httpClientImplementation;
    this.shouldTransformDates = options?.transformDates ?? true;
  }
  /**
   * List app states
   *
   * **Required scope:** state:app-states:read<br/><br/>Lists app states. By default, only provides property __key__ per state. Use *add-fields* parameter to include more fields and the *filter* parameter to narrow down the returned states.
   *
   * @returns {Promise<AppStates>} The list of app states
   *
   * @example <caption>Code example</caption>
   * import { stateClient } from "@dynatrace-sdk/client-state";
   *
   * const data = await stateClient.getAppStates();
   *
   * @throws {Unauthorized} Unauthorized
   * @throws {Forbidden} Forbidden
   * @throws {InternalServerError} Internal server error
   * @throws {GeneralError} Unexpected error
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async getAppStates(config = {}) {
    if (!config) {
      throw new ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const query = toQueryString({ "add-fields": config.addFields, filter: config.filter });
    try {
      const response = await this.httpClient.send({
        url: `/platform/state/v1/app-states${query}`,
        method: "GET",
        headers: {
          Accept: "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? transformResponse(responseValue, [["modificationInfo.lastModifiedTime", { datetime: true }]]) : responseValue;
      } catch (err) {
        throw new InvalidResponseError(
          `StateClient.getAppStates:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if (isHttpClientAbortError(e) || isInvalidResponseError(e)) {
        throw e;
      }
      if (!isHttpClientResponseError(e)) {
        throw new ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      switch (response.status) {
        case 401: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new Unauthorized(`401`, response, errorBody, getErrorMessage(errorBody, "Unauthorized"), e);
          } catch (err) {
            if (err instanceof Unauthorized) {
              throw err;
            }
            throw new InvalidResponseError(
              `StateClient.getAppStates:401`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new Forbidden(`403`, response, errorBody, getErrorMessage(errorBody, "Forbidden"), e);
          } catch (err) {
            if (err instanceof Forbidden) {
              throw err;
            }
            throw new InvalidResponseError(
              `StateClient.getAppStates:403`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 500: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new InternalServerError(
              `500`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Internal server error"),
              e
            );
          } catch (err) {
            if (err instanceof InternalServerError) {
              throw err;
            }
            throw new InvalidResponseError(
              `StateClient.getAppStates:500`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        default: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new GeneralError(`err`, response, errorBody, getErrorMessage(errorBody, "Unexpected error"), e);
          } catch (err) {
            if (err instanceof GeneralError) {
              throw err;
            }
            throw new InvalidResponseError(
              `StateClient.getAppStates:err`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
      }
    }
  }
  /**
   * Delete all app states
   *
   * **Required scope:** state:app-states:delete<br/><br/>Deletes all app states for an app to reset the app into a clean state.
   *
   * @example <caption>Code example</caption>
   * import { stateClient } from "@dynatrace-sdk/client-state";
   *
   * const data = await stateClient.deleteAppStates();
   *
   * @throws {GeneralError} Unexpected error
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async deleteAppStates(config = {}) {
    try {
      await this.httpClient.send({
        url: `/platform/state/v1/app-states`,
        method: "DELETE",
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [204].includes(status);
        }
      });
      return;
    } catch (e) {
      if (isHttpClientAbortError(e)) {
        throw e;
      }
      if (!isHttpClientResponseError(e)) {
        throw new ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      const responseValue = await response.body("json");
      try {
        const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
        throw new GeneralError(`err`, response, errorBody, getErrorMessage(errorBody, "Unexpected error"), e);
      } catch (err) {
        if (err instanceof GeneralError) {
          throw err;
        }
        throw new InvalidResponseError(
          `StateClient.deleteAppStates:err`,
          err,
          responseValue,
          "ErrorResponse",
          void 0,
          response
        );
      }
    }
  }
  /**
   * Gets app state
   *
   * **Required scope:** state:app-states:read
   *
   * @returns {Promise<AppState>} The app state
   *
   * @example <caption>Code example</caption>
   * import { stateClient } from "@dynatrace-sdk/client-state";
   *
   * const data = await stateClient.getAppState({
   *   key: "some-key",
   * });
   *
   * @throws {Unauthorized} Unauthorized
   * @throws {Forbidden} Forbidden
   * @throws {NotFound} Not found
   * @throws {InternalServerError} Internal server error
   * @throws {GeneralError} Unexpected error
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async getAppState(config) {
    try {
      const response = await this.httpClient.send({
        url: `/platform/state/v1/app-states/${config.key}`,
        method: "GET",
        headers: {
          Accept: "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? transformResponse(responseValue, [["modificationInfo.lastModifiedTime", { datetime: true }]]) : responseValue;
      } catch (err) {
        throw new InvalidResponseError(
          `StateClient.getAppState:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if (isHttpClientAbortError(e) || isInvalidResponseError(e)) {
        throw e;
      }
      if (!isHttpClientResponseError(e)) {
        throw new ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      switch (response.status) {
        case 401: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new Unauthorized(`401`, response, errorBody, getErrorMessage(errorBody, "Unauthorized"), e);
          } catch (err) {
            if (err instanceof Unauthorized) {
              throw err;
            }
            throw new InvalidResponseError(
              `StateClient.getAppState:401`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new Forbidden(`403`, response, errorBody, getErrorMessage(errorBody, "Forbidden"), e);
          } catch (err) {
            if (err instanceof Forbidden) {
              throw err;
            }
            throw new InvalidResponseError(
              `StateClient.getAppState:403`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new NotFound(`404`, response, errorBody, getErrorMessage(errorBody, "Not found"), e);
          } catch (err) {
            if (err instanceof NotFound) {
              throw err;
            }
            throw new InvalidResponseError(
              `StateClient.getAppState:404`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 500: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new InternalServerError(
              `500`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Internal server error"),
              e
            );
          } catch (err) {
            if (err instanceof InternalServerError) {
              throw err;
            }
            throw new InvalidResponseError(
              `StateClient.getAppState:500`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        default: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new GeneralError(`err`, response, errorBody, getErrorMessage(errorBody, "Unexpected error"), e);
          } catch (err) {
            if (err instanceof GeneralError) {
              throw err;
            }
            throw new InvalidResponseError(
              `StateClient.getAppState:err`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
      }
    }
  }
  /**
   * Updates app state
   *
   * **Required scope:** state:app-states:write<br/><br/>Updates the *cross-user* app state for the given key. Be aware that other users will be able to read the value.
   * Use the user-scoped [user-app-state](#/Public/setUserAppState) to only store values for the authenticated user.
   *
   * **[Certain limits](https://dt-url.net/platform-services-state-service/#limitations)** apply when updating app states.
   *
   * @example <caption>Code example</caption>
   * import { stateClient } from "@dynatrace-sdk/client-state";
   *
   * const data = await stateClient.setAppState({
   *   key: "some-key",
   *   body: { value: "some-state", validUntilTime: "now+2d" },
   * });
   *
   * @throws {AppStateLimitsExceeded} Exceeded size limit for combined size of app states of this app
   * @throws {Unauthorized} Unauthorized
   * @throws {Forbidden} Forbidden
   * @throws {InternalServerError} Internal server error
   * @throws {GeneralError} Unexpected error
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async setAppState(config) {
    if (!config) {
      throw new ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const encodedBody = this.shouldTransformDates ? transformRequest(config.body, [["modificationInfo.lastModifiedTime", { datetime: true }]]) : config.body;
    try {
      await this.httpClient.send({
        url: `/platform/state/v1/app-states/${config.key}`,
        method: "PUT",
        requestBodyType: "json",
        body: encodedBody,
        headers: {
          "Content-Type": "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      return;
    } catch (e) {
      if (isHttpClientAbortError(e)) {
        throw e;
      }
      if (!isHttpClientResponseError(e)) {
        throw new ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new AppStateLimitsExceeded(
              `400`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Exceeded size limit for combined size of app states of this app"),
              e
            );
          } catch (err) {
            if (err instanceof AppStateLimitsExceeded) {
              throw err;
            }
            throw new InvalidResponseError(
              `StateClient.setAppState:400`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 401: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new Unauthorized(`401`, response, errorBody, getErrorMessage(errorBody, "Unauthorized"), e);
          } catch (err) {
            if (err instanceof Unauthorized) {
              throw err;
            }
            throw new InvalidResponseError(
              `StateClient.setAppState:401`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new Forbidden(`403`, response, errorBody, getErrorMessage(errorBody, "Forbidden"), e);
          } catch (err) {
            if (err instanceof Forbidden) {
              throw err;
            }
            throw new InvalidResponseError(
              `StateClient.setAppState:403`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 500: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new InternalServerError(
              `500`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Internal server error"),
              e
            );
          } catch (err) {
            if (err instanceof InternalServerError) {
              throw err;
            }
            throw new InvalidResponseError(
              `StateClient.setAppState:500`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        default: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new GeneralError(`err`, response, errorBody, getErrorMessage(errorBody, "Unexpected error"), e);
          } catch (err) {
            if (err instanceof GeneralError) {
              throw err;
            }
            throw new InvalidResponseError(
              `StateClient.setAppState:err`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
      }
    }
  }
  /**
   * Deletes app state
   *
   * **Required scope:** state:app-states:delete
   *
   * @example <caption>Code example</caption>
   * import { stateClient } from "@dynatrace-sdk/client-state";
   *
   * const data = await stateClient.deleteAppState({
   *   key: "some-key",
   * });
   *
   * @throws {Unauthorized} Unauthorized
   * @throws {Forbidden} Forbidden
   * @throws {NotFound} Not found
   * @throws {InternalServerError} Internal server error
   * @throws {GeneralError} Unexpected error
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async deleteAppState(config) {
    try {
      await this.httpClient.send({
        url: `/platform/state/v1/app-states/${config.key}`,
        method: "DELETE",
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [204].includes(status);
        }
      });
      return;
    } catch (e) {
      if (isHttpClientAbortError(e)) {
        throw e;
      }
      if (!isHttpClientResponseError(e)) {
        throw new ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      switch (response.status) {
        case 401: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new Unauthorized(`401`, response, errorBody, getErrorMessage(errorBody, "Unauthorized"), e);
          } catch (err) {
            if (err instanceof Unauthorized) {
              throw err;
            }
            throw new InvalidResponseError(
              `StateClient.deleteAppState:401`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new Forbidden(`403`, response, errorBody, getErrorMessage(errorBody, "Forbidden"), e);
          } catch (err) {
            if (err instanceof Forbidden) {
              throw err;
            }
            throw new InvalidResponseError(
              `StateClient.deleteAppState:403`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new NotFound(`404`, response, errorBody, getErrorMessage(errorBody, "Not found"), e);
          } catch (err) {
            if (err instanceof NotFound) {
              throw err;
            }
            throw new InvalidResponseError(
              `StateClient.deleteAppState:404`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 500: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new InternalServerError(
              `500`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Internal server error"),
              e
            );
          } catch (err) {
            if (err instanceof InternalServerError) {
              throw err;
            }
            throw new InvalidResponseError(
              `StateClient.deleteAppState:500`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        default: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new GeneralError(`err`, response, errorBody, getErrorMessage(errorBody, "Unexpected error"), e);
          } catch (err) {
            if (err instanceof GeneralError) {
              throw err;
            }
            throw new InvalidResponseError(
              `StateClient.deleteAppState:err`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
      }
    }
  }
  /**
   * List user app states
   *
   * **Required scope:** state:user-app-states:read<br/><br/>Lists user app states. By default, only provides property __key__ per state. Use *add-fields* parameter to include more fields and the *filter* parameter to narrow down the returned states
   *
   * @returns {Promise<UserAppStates>} The list of user app states
   *
   * @example <caption>Code example</caption>
   * import { stateClient } from "@dynatrace-sdk/client-state";
   *
   * const data = await stateClient.getUserAppStates();
   *
   * @throws {Unauthorized} Unauthorized
   * @throws {Forbidden} Forbidden
   * @throws {InternalServerError} Internal server error
   * @throws {GeneralError} Unexpected error
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async getUserAppStates(config = {}) {
    if (!config) {
      throw new ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const query = toQueryString({ "add-fields": config.addFields, filter: config.filter });
    try {
      const response = await this.httpClient.send({
        url: `/platform/state/v1/user-app-states${query}`,
        method: "GET",
        headers: {
          Accept: "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? transformResponse(responseValue, [["modificationInfo.lastModifiedTime", { datetime: true }]]) : responseValue;
      } catch (err) {
        throw new InvalidResponseError(
          `StateClient.getUserAppStates:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if (isHttpClientAbortError(e) || isInvalidResponseError(e)) {
        throw e;
      }
      if (!isHttpClientResponseError(e)) {
        throw new ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      switch (response.status) {
        case 401: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new Unauthorized(`401`, response, errorBody, getErrorMessage(errorBody, "Unauthorized"), e);
          } catch (err) {
            if (err instanceof Unauthorized) {
              throw err;
            }
            throw new InvalidResponseError(
              `StateClient.getUserAppStates:401`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new Forbidden(`403`, response, errorBody, getErrorMessage(errorBody, "Forbidden"), e);
          } catch (err) {
            if (err instanceof Forbidden) {
              throw err;
            }
            throw new InvalidResponseError(
              `StateClient.getUserAppStates:403`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 500: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new InternalServerError(
              `500`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Internal server error"),
              e
            );
          } catch (err) {
            if (err instanceof InternalServerError) {
              throw err;
            }
            throw new InvalidResponseError(
              `StateClient.getUserAppStates:500`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        default: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new GeneralError(`err`, response, errorBody, getErrorMessage(errorBody, "Unexpected error"), e);
          } catch (err) {
            if (err instanceof GeneralError) {
              throw err;
            }
            throw new InvalidResponseError(
              `StateClient.getUserAppStates:err`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
      }
    }
  }
  /**
   * Delete all user app states
   *
   * **Required scope:** state:user-app-states:delete<br/><br/>Deletes all user app states for the calling user and app.
   *
   * @example <caption>Code example</caption>
   * import { stateClient } from "@dynatrace-sdk/client-state";
   *
   * const data = await stateClient.deleteUserAppStates();
   *
   * @throws {GeneralError} Unexpected error
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async deleteUserAppStates(config = {}) {
    try {
      await this.httpClient.send({
        url: `/platform/state/v1/user-app-states`,
        method: "DELETE",
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [204].includes(status);
        }
      });
      return;
    } catch (e) {
      if (isHttpClientAbortError(e)) {
        throw e;
      }
      if (!isHttpClientResponseError(e)) {
        throw new ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      const responseValue = await response.body("json");
      try {
        const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
        throw new GeneralError(`err`, response, errorBody, getErrorMessage(errorBody, "Unexpected error"), e);
      } catch (err) {
        if (err instanceof GeneralError) {
          throw err;
        }
        throw new InvalidResponseError(
          `StateClient.deleteUserAppStates:err`,
          err,
          responseValue,
          "ErrorResponse",
          void 0,
          response
        );
      }
    }
  }
  /**
   * Get user app state
   *
   * **Required scope:** state:user-app-states:read
   *
   * @returns {Promise<UserAppState>} The user app state
   *
   * @example <caption>Code example</caption>
   * import { stateClient } from "@dynatrace-sdk/client-state";
   *
   * const data = await stateClient.getUserAppState({
   *   key: "some-key",
   * });
   *
   * @throws {Unauthorized} Unauthorized
   * @throws {Forbidden} Forbidden
   * @throws {NotFound} Not found
   * @throws {InternalServerError} Internal server error
   * @throws {GeneralError} Unexpected error
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async getUserAppState(config) {
    try {
      const response = await this.httpClient.send({
        url: `/platform/state/v1/user-app-states/${config.key}`,
        method: "GET",
        headers: {
          Accept: "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? transformResponse(responseValue, [["modificationInfo.lastModifiedTime", { datetime: true }]]) : responseValue;
      } catch (err) {
        throw new InvalidResponseError(
          `StateClient.getUserAppState:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if (isHttpClientAbortError(e) || isInvalidResponseError(e)) {
        throw e;
      }
      if (!isHttpClientResponseError(e)) {
        throw new ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      switch (response.status) {
        case 401: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new Unauthorized(`401`, response, errorBody, getErrorMessage(errorBody, "Unauthorized"), e);
          } catch (err) {
            if (err instanceof Unauthorized) {
              throw err;
            }
            throw new InvalidResponseError(
              `StateClient.getUserAppState:401`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new Forbidden(`403`, response, errorBody, getErrorMessage(errorBody, "Forbidden"), e);
          } catch (err) {
            if (err instanceof Forbidden) {
              throw err;
            }
            throw new InvalidResponseError(
              `StateClient.getUserAppState:403`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new NotFound(`404`, response, errorBody, getErrorMessage(errorBody, "Not found"), e);
          } catch (err) {
            if (err instanceof NotFound) {
              throw err;
            }
            throw new InvalidResponseError(
              `StateClient.getUserAppState:404`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 500: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new InternalServerError(
              `500`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Internal server error"),
              e
            );
          } catch (err) {
            if (err instanceof InternalServerError) {
              throw err;
            }
            throw new InvalidResponseError(
              `StateClient.getUserAppState:500`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        default: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new GeneralError(`err`, response, errorBody, getErrorMessage(errorBody, "Unexpected error"), e);
          } catch (err) {
            if (err instanceof GeneralError) {
              throw err;
            }
            throw new InvalidResponseError(
              `StateClient.getUserAppState:err`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
      }
    }
  }
  /**
   * Updates user app state
   *
   * **Required scope:** state:user-app-states:write<br/><br/>Updates the user specific app state for the given key and calling user.
   *
   * **[Certain limits](https://dt-url.net/platform-services-state-service/#limitations)** apply when updating user app states.
   *
   * @example <caption>Code example</caption>
   * import { stateClient } from "@dynatrace-sdk/client-state";
   *
   * const data = await stateClient.setUserAppState({
   *   key: "some-key",
   *   body: { value: "some-state", validUntilTime: "now+2d" },
   * });
   *
   * @throws {UserAppStateLimitsExceeded} Exceeded limit for number of user app states to be stored for this user and app
   * @throws {Unauthorized} Unauthorized
   * @throws {Forbidden} Forbidden
   * @throws {InternalServerError} Internal server error
   * @throws {GeneralError} Unexpected error
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async setUserAppState(config) {
    if (!config) {
      throw new ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const encodedBody = this.shouldTransformDates ? transformRequest(config.body, [["modificationInfo.lastModifiedTime", { datetime: true }]]) : config.body;
    try {
      await this.httpClient.send({
        url: `/platform/state/v1/user-app-states/${config.key}`,
        method: "PUT",
        requestBodyType: "json",
        body: encodedBody,
        headers: {
          "Content-Type": "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      return;
    } catch (e) {
      if (isHttpClientAbortError(e)) {
        throw e;
      }
      if (!isHttpClientResponseError(e)) {
        throw new ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new UserAppStateLimitsExceeded(
              `400`,
              response,
              errorBody,
              getErrorMessage(
                errorBody,
                "Exceeded limit for number of user app states to be stored for this user and app"
              ),
              e
            );
          } catch (err) {
            if (err instanceof UserAppStateLimitsExceeded) {
              throw err;
            }
            throw new InvalidResponseError(
              `StateClient.setUserAppState:400`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 401: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new Unauthorized(`401`, response, errorBody, getErrorMessage(errorBody, "Unauthorized"), e);
          } catch (err) {
            if (err instanceof Unauthorized) {
              throw err;
            }
            throw new InvalidResponseError(
              `StateClient.setUserAppState:401`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new Forbidden(`403`, response, errorBody, getErrorMessage(errorBody, "Forbidden"), e);
          } catch (err) {
            if (err instanceof Forbidden) {
              throw err;
            }
            throw new InvalidResponseError(
              `StateClient.setUserAppState:403`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 500: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new InternalServerError(
              `500`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Internal server error"),
              e
            );
          } catch (err) {
            if (err instanceof InternalServerError) {
              throw err;
            }
            throw new InvalidResponseError(
              `StateClient.setUserAppState:500`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        default: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new GeneralError(`err`, response, errorBody, getErrorMessage(errorBody, "Unexpected error"), e);
          } catch (err) {
            if (err instanceof GeneralError) {
              throw err;
            }
            throw new InvalidResponseError(
              `StateClient.setUserAppState:err`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
      }
    }
  }
  /**
   * Delete user app state
   *
   * **Required scope:** state:user-app-states:delete
   *
   * @example <caption>Code example</caption>
   * import { stateClient } from "@dynatrace-sdk/client-state";
   *
   * const data = await stateClient.deleteUserAppState({
   *   key: "some-key",
   * });
   *
   * @throws {Unauthorized} Unauthorized
   * @throws {Forbidden} Forbidden
   * @throws {NotFound} Not found
   * @throws {InternalServerError} Internal server error
   * @throws {GeneralError} Unexpected error
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async deleteUserAppState(config) {
    try {
      await this.httpClient.send({
        url: `/platform/state/v1/user-app-states/${config.key}`,
        method: "DELETE",
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [204].includes(status);
        }
      });
      return;
    } catch (e) {
      if (isHttpClientAbortError(e)) {
        throw e;
      }
      if (!isHttpClientResponseError(e)) {
        throw new ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await apiGatewayErrorsHandler(response);
      switch (response.status) {
        case 401: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new Unauthorized(`401`, response, errorBody, getErrorMessage(errorBody, "Unauthorized"), e);
          } catch (err) {
            if (err instanceof Unauthorized) {
              throw err;
            }
            throw new InvalidResponseError(
              `StateClient.deleteUserAppState:401`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 403: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new Forbidden(`403`, response, errorBody, getErrorMessage(errorBody, "Forbidden"), e);
          } catch (err) {
            if (err instanceof Forbidden) {
              throw err;
            }
            throw new InvalidResponseError(
              `StateClient.deleteUserAppState:403`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new NotFound(`404`, response, errorBody, getErrorMessage(errorBody, "Not found"), e);
          } catch (err) {
            if (err instanceof NotFound) {
              throw err;
            }
            throw new InvalidResponseError(
              `StateClient.deleteUserAppState:404`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        case 500: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new InternalServerError(
              `500`,
              response,
              errorBody,
              getErrorMessage(errorBody, "Internal server error"),
              e
            );
          } catch (err) {
            if (err instanceof InternalServerError) {
              throw err;
            }
            throw new InvalidResponseError(
              `StateClient.deleteUserAppState:500`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
        default: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? transformResponse(responseValue, []) : responseValue;
            throw new GeneralError(`err`, response, errorBody, getErrorMessage(errorBody, "Unexpected error"), e);
          } catch (err) {
            if (err instanceof GeneralError) {
              throw err;
            }
            throw new InvalidResponseError(
              `StateClient.deleteUserAppState:err`,
              err,
              responseValue,
              "ErrorResponse",
              void 0,
              response
            );
          }
        }
      }
    }
  }
};
var stateClient = /* @__PURE__ */ new StateClient(defaultHttpClient);

// packages/client/state/src/lib/error-envelopes/index.ts
import {
  ApiClientError as ApiClientError2,
  ApiGatewayError as ApiGatewayError2,
  ClientRequestError as ClientRequestError3,
  InvalidResponseError as InvalidResponseError2,
  isApiClientError,
  isApiGatewayError,
  isClientRequestError,
  isInvalidResponseError as isInvalidResponseError2
} from "@dynatrace-sdk/shared-errors";

// packages/client/state/src/lib/models/error-details-error-code.ts
var ErrorDetailsErrorCode = /* @__PURE__ */ ((ErrorDetailsErrorCode2) => {
  ErrorDetailsErrorCode2["AppStateOverallSizeLimitExceeded"] = "AppStateOverallSizeLimitExceeded";
  ErrorDetailsErrorCode2["AppStateSizeLimitExceeded"] = "AppStateSizeLimitExceeded";
  ErrorDetailsErrorCode2["UserAppStateSizeLimitExceeded"] = "UserAppStateSizeLimitExceeded";
  ErrorDetailsErrorCode2["UserAppStateCountLimitExceeded"] = "UserAppStateCountLimitExceeded";
  ErrorDetailsErrorCode2["UserAppStateSizeLimitPerUserExceeded"] = "UserAppStateSizeLimitPerUserExceeded";
  return ErrorDetailsErrorCode2;
})(ErrorDetailsErrorCode || {});
export {
  ApiClientError2 as ApiClientError,
  ApiGatewayError2 as ApiGatewayError,
  AppStateLimitsExceeded,
  ClientRequestError3 as ClientRequestError,
  ErrorDetailsErrorCode,
  ErrorResponseError,
  Forbidden,
  GeneralError,
  InternalServerError,
  InvalidResponseError2 as InvalidResponseError,
  NotFound,
  StateClient,
  Unauthorized,
  UserAppStateLimitsExceeded,
  isApiClientError,
  isApiGatewayError,
  isAppStateLimitsExceeded,
  isClientRequestError,
  isErrorResponseError,
  isForbidden,
  isGeneralError,
  isInternalServerError,
  isInvalidResponseError2 as isInvalidResponseError,
  isNotFound,
  isUnauthorized,
  isUserAppStateLimitsExceeded,
  stateClient
};
