// @dynatrace/strato-components/esm/forms/spin-buttons/hooks/useSpinButtonInteraction.js
function useSpinButtonInteraction(props) {
  const { disabled, selectAll, deselectAll, rootElement } = props;
  return {
    onClick: (event) => {
      if (disabled) {
        event.preventDefault();
      } else {
        selectAll();
      }
    },
    onFocus: (event) => {
      if (event.target.hasAttribute("contenteditable") === false || event.target.getAttribute("contenteditable") === "false") {
        return;
      }
      const selection = window.getSelection();
      const selectionSpansMultipleNodes = selection && selection.anchorNode !== selection.focusNode && selection.toString().length > 0;
      const isRestoringSelection = event.target.closest(".strato-inputgroup")?.hasAttribute("data-restoring-selection") ?? false;
      if (selectionSpansMultipleNodes || isRestoringSelection) {
        return;
      }
      if (disabled) {
        event.preventDefault();
      } else {
        selectAll();
        rootElement?.scrollIntoView({
          block: "nearest",
          inline: "nearest"
        });
      }
    },
    onBlur: (event) => {
      if (event.target.hasAttribute("contenteditable") === false || event.target.getAttribute("contenteditable") === "false") {
        return;
      }
      deselectAll();
    }
  };
}
export {
  useSpinButtonInteraction
};
