// @dynatrace/strato-components/charts/meter-bar/single-meter-bar/components/MeterBarChartThresholds/MeterBarChartThresholds.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MeterBarChartThresholds_exports = {};
__export(MeterBarChartThresholds_exports, {
  MeterBarChartThresholds: () => MeterBarChartThresholds
});
module.exports = __toCommonJS(MeterBarChartThresholds_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_MeterBarChartThresholdIndicator = require("../MeterBarChartThresholdIndicator.js");
var import_MeterBarChartThresholds_css = require("./MeterBarChartThresholds.sty.js");
var import_constants = require("../../constants.js");
const MeterBarChartThresholds = ({
  thresholdIndicators,
  segmentHeight
}) => {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "svg",
    {
      role: "presentation",
      className: import_MeterBarChartThresholds_css.thresholdContainer,
      height: segmentHeight + import_constants.THRESHOLD_ADDITIONAL_HEIGHT,
      children: thresholdIndicators.map((thresholdIndicator) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_MeterBarChartThresholdIndicator.MeterBarChartThresholdIndicator,
        {
          position: thresholdIndicator
        },
        `indicator-${thresholdIndicator}`
      ))
    }
  );
};
