// @dynatrace/strato-components/charts/honeycomb/components/HiveOverlay.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var HiveOverlay_exports = {};
__export(HiveOverlay_exports, {
  HiveOverlay: () => HiveOverlay
});
module.exports = __toCommonJS(HiveOverlay_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_HiveOverlay_css = require("./HiveOverlay.sty.js");
var import_HoneycombToolbar = require("./toolbar/HoneycombToolbar.js");
var import_HoneycombOverlayTooltip = require("./tooltip/HoneycombOverlayTooltip.js");
var import_Flex = require("../../../layouts/flex/Flex.js");
var import_useGraphSize = require("../../core/components/chart-layout/hooks/useGraphSize.js");
var import_utils = require("../../core/components/overlay-tooltip/_to-be-removed/utils.js");
var import_OverlayContext = require("../../core/components/overlay-tooltip/contexts/OverlayContext.js");
var import_ChartSeriesActionContext = require("../../core/components/series-action/ChartSeriesActionContext.js");
var import_is_toolbar_enabled = require("../../core/utils/is-toolbar-enabled.js");
var import_HoneycombOverlayRefContext = require("../contexts/HoneycombOverlayRefContext.js");
var import_use_honeycomb_toolbar_config = require("../hooks/use-honeycomb-toolbar-config.js");
var import_use_overlay_events = require("../hooks/use-overlay-events.js");
const HiveOverlay = () => {
  const toolbarConfig = (0, import_use_honeycomb_toolbar_config.useHoneycombToolbarConfig)();
  const toolbarEnabled = (0, import_is_toolbar_enabled.isToolbarEnabled)(toolbarConfig);
  const overlay = (0, import_OverlayContext.useOverlay)();
  const {
    handlePointerEnter,
    handlePointerMove,
    handlePointerLeave,
    handleMouseClick
  } = (0, import_use_overlay_events.useOverlayEvents)(overlay);
  const { height, width } = (0, import_useGraphSize.useGraphSize)();
  const { chartSeriesActions } = (0, import_ChartSeriesActionContext.useSeriesActions)();
  const data = (0, import_utils.useTooltipData)().data;
  const series = data?.series?.[0];
  const metadata = data?.metadata;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_HoneycombOverlayRefContext.HoneycombOverlayRefContext.Provider, { value: overlay.overlayRef, children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_Flex.Flex,
    {
      className: import_HiveOverlay_css.overlay,
      style: {
        height,
        width
      },
      "data-testid": "honeycomb-overlay",
      onPointerEnter: handlePointerEnter,
      onPointerMove: handlePointerMove,
      onPointerLeave: handlePointerLeave,
      onClick: handleMouseClick,
      ref: overlay.overlayRef,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_HoneycombOverlayTooltip.HoneycombOverlayTooltip,
          {
            actionsForItem: chartSeriesActions && series && metadata ? () => () => chartSeriesActions({
              id: metadata.id,
              name: series.name,
              values: [series.value]
            }) : void 0
          }
        ),
        toolbarEnabled && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_HoneycombToolbar.HoneycombToolbar, {})
      ]
    }
  ) });
};
