// @dynatrace/strato-components/charts/honeycomb/components/HoneycombChartLayout.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var HoneycombChartLayout_exports = {};
__export(HoneycombChartLayout_exports, {
  HoneycombChartLayout: () => HoneycombChartLayout
});
module.exports = __toCommonJS(HoneycombChartLayout_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_ChartLayout = require("../../core/components/chart-layout/ChartLayout.js");
var import_useAutoLegendRefresh = require("../../core/components/chart-layout/hooks/useAutoLegendRefresh.js");
var import_get_chart_states_template = require("../../core/components/chart-states/utils/get-chart-states-template.js");
var import_constants = require("../constants.js");
var import_HoneycombChart_css = require("../HoneycombChart.sty.js");
var import_HoneycombHive = require("./HoneycombHive.js");
var import_use_honeycomb_data = require("../hooks/use-honeycomb-data.js");
var import_use_honeycomb_legend_config = require("../hooks/use-honeycomb-legend-config.js");
var import_HoneycombLegendRenderer = require("./legend/HoneycombLegendRenderer.js");
const HoneycombChartLayout = (props) => {
  const {
    data,
    width = import_constants.DEFAULT_HONEYCOMB_WIDTH,
    loading = false,
    children,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow
  } = props;
  const filteredData = (0, import_use_honeycomb_data.useHoneycombData)();
  const legendConfig = (0, import_use_honeycomb_legend_config.useHoneycombLegendConfig)();
  const legendRatio = legendConfig?.initialRatio ?? legendConfig?.ratio;
  const chartLayoutRef = (0, import_useAutoLegendRefresh.useAutoLegendRefresh)(data);
  const { emptyState, errorState } = (0, import_get_chart_states_template.getChartStatesTemplates)(children);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_ChartLayout.ChartLayout,
    {
      ref: chartLayoutRef,
      chartWidth: width,
      chartHeight: "100%",
      isEmpty: (0, import_lodash_es.isEmpty)(filteredData),
      emptyState,
      errorState,
      showLoader: loading,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ChartLayout.ChartLayout.Graph, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: import_HoneycombChart_css.graphContainer, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_HoneycombHive.HoneycombHive, {}) }) }),
        !(0, import_lodash_es.isUndefined)(legendConfig) && !legendConfig.hidden && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_ChartLayout.ChartLayout.Legend,
          {
            position: legendConfig.position,
            initialRatio: legendRatio,
            onResize: legendConfig.onRatioChange,
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_HoneycombLegendRenderer.HoneycombLegendRenderer, {})
          }
        )
      ]
    }
  );
};
HoneycombChartLayout["displayName"] = "HoneycombChartLayout";
