// @dynatrace-sdk/error-handlers/esm/index.js
/**
 * @license
 * Copyright 2023 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// packages/platform/shared/utils/get-global-with-dt-runtime.ts
var getGlobalWithDtRuntime = () => typeof globalThis !== "undefined" ? globalThis : window;

// packages/platform/error-handlers/src/lib/add-global-error-serializer.ts
function addGlobalErrorSerializer(serializer) {
  const addGlobalErrorSerializerFunc = getAddGlobalErrorSerializer();
  if (typeof addGlobalErrorSerializerFunc === "function") {
    addGlobalErrorSerializerFunc(serializer);
  } else {
    console.warn("Missing addGlobalErrorSerializer function from web runtime.");
  }
}
function isGlobalErrorSerializerSupported() {
  return typeof getAddGlobalErrorSerializer() === "function";
}
function getAddGlobalErrorSerializer() {
  return getGlobalWithDtRuntime().dtRuntime?.errorHandlers?.addGlobalErrorSerializer;
}

// packages/platform/error-handlers/src/lib/types/error-type.ts
var ErrorType = /* @__PURE__ */ ((ErrorType2) => {
  ErrorType2["COMMON"] = "JS Error";
  ErrorType2["HTTP"] = "Http Error";
  return ErrorType2;
})(ErrorType || {});
export {
  ErrorType,
  addGlobalErrorSerializer,
  isGlobalErrorSerializerSupported
};
