// @dynatrace/strato-components/tables/DataTable/features/RowInteractivity/RowInteractivity.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var RowInteractivity_exports = {};
__export(RowInteractivity_exports, {
  DataTableRowInteractivity: () => DataTableRowInteractivity,
  useRowInteractivity: () => useRowInteractivity,
  useRowInteractivityProps: () => useRowInteractivityProps
});
module.exports = __toCommonJS(RowInteractivity_exports);
var import_react_table = require("@tanstack/react-table");
var import_react = require("react");
var import_react_intl = require("react-intl");
var styles = __toESM(require("./RowInteractivity.sty.js"));
var import_useControllableState = require("../../../../core/hooks/useControllableState.js");
var import_useFocusRing = require("../../../../core/styles/useFocusRing.js");
var import_merge_props = require("../../../../core/utils/merge-props.js");
var import_is_target_interactive = require("../../../utils/is-target-interactive.js");
var import_constants = require("../../constants.js");
var import_DataTableContext = require("../../contexts/DataTableContext.js");
var import_utils = require("../../hooks/useTable/utils.js");
var import_messages = require("../../messages.js");
function removeHoverStyles(virtualizationContainer) {
  if (virtualizationContainer) {
    const hoveredRows = virtualizationContainer.querySelectorAll(
      `.${styles.interactiveRow.classNames.variants.hovered.true}`
    );
    for (const hoveredRow of Array.from(hoveredRows)) {
      hoveredRow.classList.remove(
        styles.interactiveRow.classNames.variants.hovered.true
      );
    }
  }
}
function addHoverStyles(row) {
  row.classList.add(styles.interactiveRow.classNames.variants.hovered.true);
}
function applyHoverStylesByRowId(virtualizationContainer, rowId) {
  if (!virtualizationContainer) {
    return;
  }
  const rowElement = virtualizationContainer.querySelector(
    `[data-rowid="${CSS.escape(rowId)}"]`
  );
  if (rowElement) {
    addHoverStyles(rowElement);
  }
}
function maintainOpenMenuRowHover(table, virtualizationContainer) {
  const openMenuRowId = table.getOpenMenuRowId();
  if (openMenuRowId === null) {
    return false;
  }
  if (table.getHoveredRow() !== openMenuRowId) {
    table.setHoveredRow(openMenuRowId);
    removeHoverStyles(virtualizationContainer);
    applyHoverStylesByRowId(virtualizationContainer, openMenuRowId);
  }
  return true;
}
function useRowInteractivityProps(row, scrollToIndex) {
  const table = (0, import_DataTableContext.useDataTable)();
  const intl = (0, import_react_intl.useIntl)();
  const clicked = (0, import_react.useRef)(false);
  const { isFocusVisible, focusProps } = (0, import_useFocusRing.useFocusRing)();
  if (!row.getCanActivateRow()) {
    return { rowInteractivityClassName: "", rowInteractivityProps: void 0 };
  }
  const isRowDragging = table.getState().draggingRowInfo !== void 0;
  const isActivated = row.getIsActivated();
  const isAutoActivationEnabled = table.options.enableAutoActivation;
  const rowInteractivityClassName = styles.interactiveRow({
    hovered: table.getHoveredRow() === row.id,
    active: isActivated,
    focus: isFocusVisible,
    contained: table.getVisualVariantOptions().contained
  });
  function handleRowActivation(event) {
    const { target } = event;
    if (table.options.enableRowInteractivity !== true || target instanceof HTMLElement === false && target instanceof SVGElement === false || // prevent deactivation if the user clicks on a Gantt segment (SVG <rect>) and the row is already active
    target.nodeName === "rect" && isActivated) {
      return;
    }
    const rowElement = target.closest('[role="row"]');
    if (rowElement && !(0, import_is_target_interactive.isTargetInteractive)(rowElement, target)) {
      row.toggleActivated(event);
    }
  }
  function isSelectionInsideTable() {
    const selection = globalThis.getSelection();
    if (!selection?.toString()) {
      return false;
    }
    const container = table.getRowVirtualizationContainer();
    if (!container) {
      return false;
    }
    let anchorNode = selection.anchorNode;
    if (anchorNode?.nodeType === Node.TEXT_NODE) {
      anchorNode = anchorNode.parentElement;
    }
    return !!anchorNode && container.contains(anchorNode);
  }
  const rowInteractivityProps = {
    tabIndex: 0,
    "data-trigger": "interactive-row",
    "aria-label": intl.formatMessage(import_messages.dataTableMessages.interactiveRowTrigger, {
      isActivated,
      rowId: row.id
    }),
    onPointerDown: (event) => {
      clicked.current = true;
    },
    onClick: (event) => {
      if (!isSelectionInsideTable() && event.detail === 1) {
        handleRowActivation(event);
      }
    },
    onAuxClick: (event) => {
      if (event.button === 1) {
        handleRowActivation(event);
      }
    },
    onKeyDown: (event) => {
      if (isRowDragging) {
        return;
      }
      const { key, target } = event;
      if (key === "Enter" && !isAutoActivationEnabled && (target instanceof HTMLElement || target instanceof SVGElement)) {
        const rowElement = target.closest('[role="row"]');
        if (rowElement && !(0, import_is_target_interactive.isTargetInteractive)(rowElement, target)) {
          event.preventDefault();
          event.stopPropagation();
          row.toggleActivated(event);
          return;
        }
      }
      let offset = null;
      switch (key) {
        case "ArrowDown": {
          offset = 1;
          break;
        }
        case "ArrowUp": {
          offset = -1;
          break;
        }
        case "PageUp": {
          offset = -10;
          break;
        }
        case "PageDown": {
          offset = 10;
          break;
        }
        case "Home": {
          offset = -Infinity;
          break;
        }
        case "End": {
          offset = Infinity;
          break;
        }
      }
      if (offset === null) {
        return;
      }
      event.preventDefault();
      event.stopPropagation();
      if (!(event.target instanceof HTMLElement)) {
        return;
      }
      const allRows = table.getRowModel().rows;
      const currentRowIndex = allRows.findIndex(
        (currentRow) => currentRow.id === row.id
      );
      if (currentRowIndex === -1) {
        return;
      }
      const clampedIndex = Math.max(
        0,
        Math.min(currentRowIndex + offset, allRows.length - 1)
      );
      const nextRow = allRows.at(clampedIndex);
      if (nextRow === void 0) {
        return;
      }
      scrollToIndex(clampedIndex, {
        align: "auto"
      });
      requestAnimationFrame(() => {
        const nextRowButton = table.getRowVirtualizationContainer()?.querySelector(
          `[data-rowid="${CSS.escape(
            nextRow.id
          )}"][data-trigger="interactive-row"]`
        );
        nextRowButton?.focus();
      });
    },
    ...(0, import_merge_props.mergeProps)(focusProps, {
      onFocus: (event) => {
        if (event.currentTarget !== event.target) {
          return;
        }
        if (clicked.current) {
          clicked.current = false;
          return;
        }
        if (!isActivated && isAutoActivationEnabled) {
          row.toggleActivated(event);
        }
      }
    })
  };
  return { rowInteractivityClassName, rowInteractivityProps };
}
function getRowInfoFromMouseEvent(event) {
  const { target } = event;
  if (!(target instanceof HTMLElement) && !(target instanceof SVGElement)) {
    return null;
  }
  const row = target.closest('[role="row"]');
  const rowId = row instanceof HTMLElement && "rowid" in row.dataset ? row.dataset.rowid : void 0;
  const isRowDetailsRow = rowId?.endsWith(import_constants.DATA_TABLE_ROW_EXPANDED_ID_SUFFIX) ?? false;
  return { target, row, rowId, isRowDetailsRow };
}
const DataTableRowInteractivity = {
  createTable(table) {
    let hoveredRow = null;
    table.setActiveRow = (updater, event) => {
      table.options.onActiveRowChange?.(updater, event);
    };
    table.setHoveredRow = (rowId) => {
      if (rowId !== hoveredRow) {
        hoveredRow = rowId;
      }
    };
    table.getHoveredRow = () => {
      return hoveredRow;
    };
    table.handleInteractiveRowMouseOver = (event) => {
      if (table.options.enableRowInteractivity !== true) {
        return;
      }
      const rowInfo = getRowInfoFromMouseEvent(event);
      if (!rowInfo) {
        return;
      }
      const { target, row, rowId, isRowDetailsRow } = rowInfo;
      const virtualizationContainer = table.getRowVirtualizationContainer();
      if (!virtualizationContainer?.contains(target) || row === null || isRowDetailsRow) {
        if (target instanceof HTMLElement && target.dataset.trigger === "context-action") {
          return;
        }
        if (maintainOpenMenuRowHover(table, virtualizationContainer)) {
          return;
        }
        table.setHoveredRow(null);
        removeHoverStyles(virtualizationContainer);
        return;
      }
      if (row && rowId && !isRowDetailsRow && table.getHoveredRow() !== rowId) {
        table.setHoveredRow(rowId);
        removeHoverStyles(virtualizationContainer);
        addHoverStyles(row);
      }
    };
    table.handleInteractiveRowMouseLeave = (event) => {
      if (table.options.enableRowInteractivity !== true) {
        return;
      }
      const virtualizationContainer = table.getRowVirtualizationContainer();
      if (maintainOpenMenuRowHover(table, virtualizationContainer)) {
        return;
      }
      table.setHoveredRow(null);
      removeHoverStyles(virtualizationContainer);
    };
    table.handleInteractiveRowClick = (event) => {
      if (table.options.enableRowInteractivity !== true) {
        return;
      }
      const rowInfo = getRowInfoFromMouseEvent(event);
      if (!rowInfo) {
        return;
      }
      const { target, row, rowId, isRowDetailsRow } = rowInfo;
      if (row && rowId && !isRowDetailsRow && !(0, import_is_target_interactive.isTargetInteractive)(row, target)) {
        event.stopPropagation();
        row.focus();
      }
    };
    (0, import_utils.assignRowMethods)(table, {
      /**
       * Returns a boolean value if the row can be activated.
       */
      getCanActivateRow() {
        return table.options.enableRowInteractivity === true;
      },
      /**
       * Returns a boolean value if the row is currently activated.
       */
      getIsActivated() {
        return table.getState().activeRow === this.id;
      },
      /**
       * Activates the row, if it is not activated.
       * Deactivates the row, if it is currently activated.
       */
      toggleActivated(event) {
        table.setActiveRow(this.getIsActivated() ? null : this.id, event);
      },
      /**
       * DEPRECATION: [APPDEV-15075] BREAKING-CHANGE: Remove rowInteractivity link
       */
      getLinkForRow() {
        return table.options.getLinkForRow?.(this);
      }
    });
  }
};
function useRowInteractivity(props, options) {
  const {
    activeRow: activeRowProp,
    defaultActiveRow,
    onActiveRowChange,
    interactiveRows
  } = props;
  const [activeRow, setActiveRow] = (0, import_useControllableState.useControllableState)({
    value: activeRowProp,
    defaultValue: defaultActiveRow
  });
  options.enableRowInteractivity = interactiveRows !== void 0 && interactiveRows !== false;
  const getLinkForRow = (0, import_react.useCallback)(
    (row) => {
      if (interactiveRows !== void 0 && typeof interactiveRows === "object" && // eslint-disable-next-line @typescript-eslint/no-deprecated
      interactiveRows.link) {
        return interactiveRows.link(row.original);
      }
      return void 0;
    },
    [interactiveRows]
  );
  if (options.enableRowInteractivity) {
    options.onActiveRowChange = (updater, event) => {
      const nextState = (0, import_react_table.functionalUpdate)(
        updater,
        options.state?.activeRow ?? null
      );
      setActiveRow(nextState);
      onActiveRowChange?.(nextState, event);
    };
    options.enableAutoActivation = interactiveRows && (interactiveRows === true || interactiveRows.autoActivate !== false);
    options.getLinkForRow = getLinkForRow;
  }
  const firstRun = (0, import_react.useRef)(true);
  (0, import_react.useEffect)(
    () => {
      if (firstRun.current) {
        firstRun.current = false;
        return;
      }
      if (options.enableRowInteractivity) {
        options.onActiveRowChange?.(null);
      }
    },
    // We only want to listen to the individual options parts and not the entire options.
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [options.enableRowInteractivity, options.data]
  );
  options.state = {
    ...options.state,
    activeRow: activeRow ?? null
  };
}
