// @dynatrace/strato-components/esm/charts/core/components/threshold-bar/markers/CircleMarker.js
import { jsx, jsxs } from "react/jsx-runtime";
import { Group } from "@visx/group";
import { useRef } from "react";
import {
  MARKER_STROKE_WIDTH,
  MARKER_STROKE_COLOR,
  MARKER_RADIUS,
  MARKER_OPACITY
} from "./constants.js";
import { useId } from "../../../../../core/hooks/useId.js";
import { useMouseHover } from "../../../hooks/useMouseHover.js";
import { ThresholdBrightnessFilter } from "../ThresholdBrightnessFilter.js";
const CircleMarker = (props) => {
  const { x, y, color, onMouseMove, onMouseLeave } = props;
  const circleRef = useRef(null);
  const hovered = useMouseHover(circleRef);
  const id = useId();
  const brightnessFilterId = `threshold-circle-marker-filter-${id}`;
  return /* @__PURE__ */ jsxs(Group, { children: [
    /* @__PURE__ */ jsx(ThresholdBrightnessFilter, { id: brightnessFilterId }),
    /* @__PURE__ */ jsx(
      "circle",
      {
        role: "graphics-symbol",
        "aria-label": "circle marker",
        ref: circleRef,
        opacity: MARKER_OPACITY,
        stroke: MARKER_STROKE_COLOR,
        strokeWidth: MARKER_STROKE_WIDTH,
        cx: x,
        cy: y,
        r: MARKER_RADIUS,
        fill: color,
        onMouseMove,
        onMouseLeave,
        filter: hovered ? `url(#${brightnessFilterId})` : void 0
      }
    )
  ] });
};
export {
  CircleMarker
};
