// @dynatrace/strato-components/esm/charts/core/components/threshold-bar/markers/PillMarker.js
import { jsx, jsxs } from "react/jsx-runtime";
import { Group } from "@visx/group";
import { useRef } from "react";
import {
  MARKER_STROKE_WIDTH,
  MARKER_STROKE_COLOR,
  MARKER_RADIUS
} from "./constants.js";
import { useId } from "../../../../../core/hooks/useId.js";
import { useMouseHover } from "../../../hooks/useMouseHover.js";
import { ThresholdBrightnessFilter } from "../ThresholdBrightnessFilter.js";
const PillMarker = (props) => {
  const { x, y0, y1, color, onMouseMove, onMouseLeave, opacity } = props;
  const rectRef = useRef(null);
  const hovered = useMouseHover(rectRef);
  const id = useId();
  const brightnessFilterId = `threshold-pill-marker-filter-${id}`;
  return /* @__PURE__ */ jsxs(Group, { children: [
    /* @__PURE__ */ jsx(ThresholdBrightnessFilter, { id: brightnessFilterId }),
    /* @__PURE__ */ jsx(
      "rect",
      {
        role: "graphics-symbol",
        "aria-label": "pill marker",
        ref: rectRef,
        x,
        y: Math.abs(y1 - y0) > MARKER_RADIUS * 2 ? Math.min(y0, y1) : (y0 + y1) / 2 - MARKER_RADIUS,
        height: Math.max(Math.abs(y1 - y0), MARKER_RADIUS * 2),
        width: MARKER_RADIUS * 2,
        fill: color,
        rx: MARKER_RADIUS,
        opacity,
        filter: hovered ? `url(#${brightnessFilterId})` : void 0,
        stroke: MARKER_STROKE_COLOR,
        strokeWidth: MARKER_STROKE_WIDTH,
        onMouseMove,
        onMouseLeave
      }
    )
  ] });
};
export {
  PillMarker
};
