// @dynatrace/strato-components/esm/tables/DataTable/features/MeterBarChart/MeterBarChart.js
import { jsx, jsxs } from "react/jsx-runtime";
import { isEmpty, isNumber } from "lodash-es";
import { useEffect, useMemo, useRef } from "react";
import Spacings from "@dynatrace/strato-design-tokens/spacings";
import { MultiMeterBarChart } from "../../../../charts/meter-bar/multi-meter-bar/MultiMeterBarChart.js";
import { METER_BAR_SIZE } from "../../../../charts/meter-bar/shared/constants.js";
import { MeterBarChart } from "../../../../charts/meter-bar/single-meter-bar/MeterBarChart.js";
import { devLog } from "../../../../core/utils/logging.js";
import { useDataTable } from "../../contexts/DataTableContext.js";
import { isDisplayColumnDef } from "../../hooks/useTable/datatable-typeguards.js";
const DEFAULT_MAX_VALUE = 100;
function isDataTableMeterBarColumnConfig(config) {
  return config !== null && config !== void 0 && typeof config === "object" && ("max" in config || "color" in config || "threshold" in config || "showTooltip" in config || "max" in config);
}
function isColumnDefWithAccessor(column) {
  return column !== null && column !== void 0 && typeof column === "object" && "accessor" in column;
}
function isMultiMeterBarData(value) {
  return Array.isArray(value) && value.every(
    (item) => typeof item === "object" && typeof item.name === "string" && typeof item.value === "number" && (item.color === void 0 || typeof item.color === "string")
  );
}
function MeterBarChartRenderer({
  value,
  column
}) {
  const table = useDataTable();
  const chartRef = useRef(null);
  let config = {};
  if (column?.columnDef.meta && isDataTableMeterBarColumnConfig(column.columnDef.meta.config)) {
    config = column.columnDef.meta.config;
  }
  const columnId = column.columnDef.id ?? "";
  const maxValuesFromState = table.getState().meterbarChartMaxValues;
  const maxValue = !isEmpty(maxValuesFromState) ? maxValuesFromState[columnId] : DEFAULT_MAX_VALUE;
  const listenerRef = useRef(null);
  useEffect(() => {
    return () => {
      listenerRef.current?.();
    };
  }, []);
  const registerUnpinOnPointerUp = config?.showTooltip ? () => {
    if (listenerRef.current) {
      return;
    }
    const handler = (event) => {
      if (event.target.closest('[role="tooltip"]') !== null) {
        return;
      }
      chartRef.current?.unpinTooltip();
      listenerRef.current?.();
      listenerRef.current = null;
    };
    listenerRef.current = () => document.removeEventListener("pointerup", handler);
    document.addEventListener("pointerup", handler);
  } : void 0;
  if (typeof value === "number") {
    return /* @__PURE__ */ jsx(
      "span",
      {
        style: { display: "contents" },
        onClick: registerUnpinOnPointerUp,
        role: "none",
        children: /* @__PURE__ */ jsxs(
          MeterBarChart,
          {
            value,
            min: config?.min,
            max: maxValue,
            color: config?.color,
            size: METER_BAR_SIZE.Size8,
            width: "100%",
            height: "auto",
            style: {
              pointerEvents: "auto",
              padding: `${Spacings.Size6} 0`
            },
            formatter: config?.formatter,
            ref: chartRef,
            children: [
              /* @__PURE__ */ jsx(MeterBarChart.Tooltip, { hidden: !config?.showTooltip }),
              config?.thresholds?.map((threshold) => /* @__PURE__ */ jsx(
                MeterBarChart.ThresholdIndicator,
                {
                  name: threshold.name,
                  color: threshold.color,
                  value: threshold.value,
                  showIndicator: threshold.showIndicator
                },
                `threshold-${threshold.color}-${threshold.value}`
              ))
            ]
          }
        )
      }
    );
  }
  const isMultiMeterData = isMultiMeterBarData(value);
  if (!isMultiMeterData) {
    devLog("Invalid meter chart data.");
  }
  return /* @__PURE__ */ jsx(
    "span",
    {
      style: { display: "contents" },
      onClick: registerUnpinOnPointerUp,
      role: "none",
      children: /* @__PURE__ */ jsxs(
        MultiMeterBarChart,
        {
          size: "size8",
          min: config?.min,
          max: maxValue,
          colorPalette: config?.colorPalette || "categorical",
          width: "100%",
          height: "auto",
          style: {
            pointerEvents: "auto",
            padding: `${Spacings.Size6} 0`
          },
          formatter: config?.formatter,
          ref: chartRef,
          // only if this is multi meter data render segments
          children: [
            isMultiMeterData && value.map((data) => /* @__PURE__ */ jsx(
              MultiMeterBarChart.Segment,
              {
                name: data.name,
                value: data.value,
                color: data?.color
              },
              `${data.name}-${data.value}`
            )),
            /* @__PURE__ */ jsx(
              MultiMeterBarChart.Tooltip,
              {
                variant: "shared",
                hidden: !config?.showTooltip
              }
            )
          ]
        }
      )
    }
  );
}
function getTotalMaxFromMeterbarChartData(data, columns) {
  const meterbarColumns = columns.filter(
    (columnDef) => isDisplayColumnDef(columnDef) && columnDef.columnType === "meterbar"
  );
  const maxValuesStorage = {};
  for (const row of data) {
    for (const columnDef of meterbarColumns) {
      if (isColumnDefWithAccessor(columnDef)) {
        const { accessor, config, id: columnId } = columnDef;
        const configuredMax = isDataTableMeterBarColumnConfig(config) ? config.max : void 0;
        const meterbarValue = row;
        const meterbarDataEntry = typeof accessor === "string" ? meterbarValue[accessor] : accessor(row);
        let totalValue = DEFAULT_MAX_VALUE;
        if (isNumber(configuredMax)) {
          maxValuesStorage[columnId] = configuredMax;
          continue;
        }
        if (meterbarDataEntry) {
          if (Array.isArray(meterbarDataEntry)) {
            totalValue = meterbarDataEntry.reduce(
              (acc, current) => acc + current.value,
              0
            );
          } else {
            const entry = isNumber(meterbarDataEntry) ? meterbarDataEntry : DEFAULT_MAX_VALUE;
            totalValue = configuredMax === "data-max" ? entry : Math.max(entry, DEFAULT_MAX_VALUE);
          }
        }
        maxValuesStorage[columnId] = Math.max(
          maxValuesStorage[columnId] ?? totalValue,
          totalValue
        );
      }
    }
  }
  return maxValuesStorage;
}
function useMeterBarChartMaxValuesCalc(props, options) {
  const meterbarChartMaxValues = useMemo(
    () => getTotalMaxFromMeterbarChartData(props.data, props.columns),
    [props.columns, props.data]
  );
  options.state = { ...options.state, meterbarChartMaxValues };
}
export {
  MeterBarChartRenderer,
  getTotalMaxFromMeterbarChartData,
  isMultiMeterBarData,
  useMeterBarChartMaxValuesCalc
};
