// @dynatrace/strato-components/esm/core/components/time-indicator/TimeIndicator.js
import { jsx } from "react/jsx-runtime";
import { TZDateMini } from "@date-fns/tz";
import { FormattedMessage, FormattedRelativeTime } from "react-intl";
import { getTimezone } from "@dynatrace-sdk/user-preferences";
import { useIntervalCallback } from "../../time/useIntervalCallback.js";
const MilliSecondsInfo = {
  OneMinute: 6e4,
  OneHour: 36e5,
  OneDay: 864e5
};
const TimeIndicator = (props) => {
  const {
    time,
    "data-testid": dataTestId,
    className: consumerClassName,
    style: consumerStyle,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    ...remainingProps
  } = props;
  const date = TZDateMini.tz(getTimezone());
  const { value: now } = useIntervalCallback(
    () => date.getTime(),
    [],
    (time - date.getSeconds() * 1e3) % MilliSecondsInfo.OneMinute,
    MilliSecondsInfo.OneMinute
  );
  const currentTime = now ?? date.getTime();
  const diff = currentTime - time;
  let content;
  if (diff < MilliSecondsInfo.OneMinute) {
    content = /* @__PURE__ */ jsx(
      FormattedMessage,
      {
        defaultMessage: "Just now",
        id: "HPbv8ZQrbBL7hRCT",
        description: "Just now marker in the FilterField recent suggestions"
      }
    );
  } else if (diff < MilliSecondsInfo.OneHour) {
    const minutes = -Math.floor(diff / MilliSecondsInfo.OneMinute);
    content = // eslint rule is not smart in checking this. The style prop comes from the lib.Intl types, where it defines the style of the formatted value.
    // eslint-disable-next-line react/style-prop-object
    /* @__PURE__ */ jsx(FormattedRelativeTime, { value: minutes, unit: "minute", style: "narrow" });
  } else if (diff < MilliSecondsInfo.OneDay) {
    const hours = -Math.floor(diff / MilliSecondsInfo.OneHour);
    content = // eslint rule is not smart in checking this. The style prop comes from the lib.Intl types, where it defines the style of the formatted value.
    // eslint-disable-next-line react/style-prop-object
    /* @__PURE__ */ jsx(FormattedRelativeTime, { value: hours, unit: "hour", style: "narrow" });
  } else {
    const days = -Math.floor(diff / MilliSecondsInfo.OneDay);
    content = /* @__PURE__ */ jsx(FormattedRelativeTime, { value: days, unit: "day", style: "narrow" });
  }
  return /* @__PURE__ */ jsx(
    "span",
    {
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...remainingProps,
      children: content
    }
  );
};
export {
  TimeIndicator
};
