// @dynatrace/strato-components/esm/content/progress/ProgressBar.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import {
  forwardRef,
  Children
} from "react";
import { isElement } from "react-is";
import { SharedProgressBarPropsContext } from "./contexts/SharedProgressBarPropsContext.js";
import { useProgressAriaProps } from "./hooks/useProgressAriaProps.js";
import {
  progressBarFillCSS,
  progressCSS,
  progressRoot
} from "./ProgressBar.sty.js";
import { ProgressBarIcon } from "./ProgressBarIcon.js";
import { ProgressBarLabel } from "./ProgressBarLabel.js";
import { ProgressBarValue } from "./ProgressBarValue.js";
import { useAriaLabelingProps } from "../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useId } from "../../core/hooks/useId.js";
import { useSafeForwardProps } from "../../core/hooks/useSafeForwardProps.js";
import { useContainerColoring } from "../../layouts/container/Container.js";
import { Flex } from "../../layouts/flex/Flex.js";
import { containerColorsCSS } from "../../styles/container.sty.js";
const COMPONENT_NAME = "ProgressBar";
function getProgressBarSlots(children) {
  const slots = {
    value: null,
    label: null,
    icon: null
  };
  Children.forEach(children, (child) => {
    if (isElement(child)) {
      switch (child.type) {
        case ProgressBarLabel:
          slots.label = child;
          break;
        case ProgressBarIcon:
          slots.icon = child;
          break;
        case ProgressBarValue:
        default:
          slots.value = child;
          break;
      }
    } else {
      if (!slots.value) {
        slots.value = child;
      }
    }
  });
  return slots;
}
const _ProgressBar = forwardRef(
  (props, forwardedRef) => {
    const {
      value: originalValue,
      max = 100,
      min = 0,
      density = "default",
      color,
      children,
      id: propId,
      "aria-valuetext": ariaValuetext,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...remainingProps
    } = props;
    const containerColoring = useContainerColoring();
    const fallbackColor = containerColoring.default ? "primary" : containerColoring.color;
    const derivedColor = color ?? fallbackColor;
    const onAccent = containerColoring.variant === "accent";
    const slots = getProgressBarSlots(children);
    const { value, isIndeterminate, progressAriaProps } = useProgressAriaProps(
      originalValue,
      min,
      max,
      ariaValuetext
    );
    const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
    const ariaLabelingProps = useAriaLabelingProps(restProps);
    const forwardProps = useSafeForwardProps(restProps, ariaLabelingProps);
    const generatedId = useId();
    const id = propId ?? generatedId;
    if (!slots.label && !ariaLabelingProps["aria-label"] && !ariaLabelingProps["aria-labelledby"]) {
      // eslint-disable-next-line no-console
      console.warn(
        "If no label is provided please make sure to provide an aria label on your progressCircle element."
      );
    }
    const progressStyles = {};
    if (!isIndeterminate && value !== void 0) {
      const percentage = (value - min) / (max - min);
      progressStyles["width"] = `${Math.round(percentage * 100)}%`;
    }
    return /* @__PURE__ */ jsx(SharedProgressBarPropsContext.Provider, { value: { id, density }, children: /* @__PURE__ */ jsxs(
      Flex,
      {
        ref: forwardedRef,
        flexDirection: "column",
        gap: 6,
        width: "100%",
        className: clsx(
          containerColorsCSS({
            color: derivedColor,
            variant: containerColoring.variant
          }),
          progressRoot({
            onAccent,
            color: derivedColor
          }),
          consumerClassName
        ),
        style: consumerStyle,
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        ...behavioralTrackingProps,
        ...forwardProps,
        "data-dt-component": COMPONENT_NAME,
        children: [
          /* @__PURE__ */ jsxs(Flex, { justifyContent: "space-between", gap: 8, children: [
            density === "condensed" ? /* @__PURE__ */ jsxs(
              Flex,
              {
                justifyContent: "space-between",
                alignItems: "flex-end",
                flex: 1,
                gap: 16,
                minWidth: 0,
                children: [
                  slots.label,
                  slots.value
                ]
              }
            ) : slots.label,
            slots.icon
          ] }),
          /* @__PURE__ */ jsx(
            "div",
            {
              className: progressCSS({
                size: density
              }),
              id,
              ...progressAriaProps,
              ...ariaLabelingProps,
              children: /* @__PURE__ */ jsx(
                "div",
                {
                  className: clsx(
                    progressBarFillCSS({
                      indeterminate: isIndeterminate
                    })
                  ),
                  style: progressStyles
                }
              )
            }
          ),
          density === "default" && slots.value
        ]
      }
    ) });
  }
);
_ProgressBar.displayName = COMPONENT_NAME;
const ProgressBar = Object.assign(_ProgressBar, {
  Value: ProgressBarValue,
  Label: ProgressBarLabel,
  Icon: ProgressBarIcon
});
export {
  ProgressBar
};
