// @dynatrace/strato-components/esm/content/progress/ProgressBarValue.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import { forwardRef } from "react";
import { useIntl } from "react-intl";
import { useSharedProgressBarPropsContext } from "./hooks/useProgressBarPropsContext.js";
import { value } from "./ProgressBarValue.sty.js";
import { canReceiveDefaultTextFormat } from "../../core/utils/can-receive-default-text-format.js";
import { textStyle } from "../../styles/textStyle.sty.js";
const ProgressBarValue = forwardRef((props, forwardedRef) => {
  const {
    children,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    ...remainingProps
  } = props;
  const intl = useIntl();
  const progressBarDefaultLabel = intl.formatMessage({
    defaultMessage: "Loading",
    id: "3J6qHQEUwDxvZuIl",
    description: "Default label for the progress-bar."
  });
  const { density } = useSharedProgressBarPropsContext();
  const childIsText = canReceiveDefaultTextFormat(children);
  return /* @__PURE__ */ jsx(
    "div",
    {
      "data-testid": dataTestId,
      ref: forwardedRef,
      className: clsx(
        consumerClassName,
        value({ size: density, childIsText }),
        textStyle({ textStyle: "small" })
      ),
      style: consumerStyle,
      ...remainingProps,
      children: children || progressBarDefaultLabel
    }
  );
});
ProgressBarValue.displayName = "ProgressBar.Value";
export {
  ProgressBarValue
};
