// @dynatrace/strato-components/esm/content/progress/ProgressCircle.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import { forwardRef } from "react";
import { useProgressAriaProps } from "./hooks/useProgressAriaProps.js";
import {
  progressCircleBackgroundCSS,
  progressCircleContainerCSS,
  progressCircleCSS,
  progressCircleRootCSS,
  progressCircleSvgCss
} from "./ProgressCircle.sty.js";
import { useAriaLabelingProps } from "../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useId } from "../../core/hooks/useId.js";
import { useSafeForwardProps } from "../../core/hooks/useSafeForwardProps.js";
import { useContainerColoring } from "../../layouts/container/Container.js";
import { containerColorsCSS } from "../../styles/container.sty.js";
import { textStyle } from "../../styles/textStyle.sty.js";
const COMPONENT_NAME = "ProgressCircle";
const progressCircleSizes = {
  small: 20,
  large: 40
};
const progressCircleThickness = 5;
const ProgressCircle = forwardRef(
  (props, forwardedRef) => {
    const {
      value: originalValue,
      min = 0,
      max = 100,
      size = "large",
      color,
      children,
      id: propId,
      "aria-valuetext": ariaValuetext,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
    const { value, isIndeterminate, progressAriaProps } = useProgressAriaProps(
      originalValue,
      min,
      max,
      ariaValuetext
    );
    const containerColoring = useContainerColoring();
    const derivedVariant = containerColoring.variant === "accent" ? "accent" : "default";
    const fallbackColor = containerColoring.default ? "primary" : containerColoring.color;
    const derivedColor = color ?? fallbackColor;
    const ariaLabelingProps = useAriaLabelingProps(restProps);
    const forwardProps = useSafeForwardProps(restProps, ariaLabelingProps);
    const generatedId = useId();
    const id = propId ?? generatedId;
    const circleStyles = {};
    if (!isIndeterminate && value !== void 0) {
      const circumference = 2 * Math.PI * ((progressCircleSizes[size] - progressCircleThickness) / 2);
      circleStyles["strokeDasharray"] = circumference.toFixed(3);
      const dashOffset = Number(
        ((max - value) / max * circumference).toFixed(3)
      );
      circleStyles["strokeDashoffset"] = `${dashOffset >= 0 ? dashOffset : 0}px`;
    }
    if (!children && !ariaLabelingProps["aria-label"] && !ariaLabelingProps["aria-labelledby"]) {
      // eslint-disable-next-line no-console
      console.warn(
        "If no label is provided please make sure to provide an aria label on your progressCircle element."
      );
    }
    return /* @__PURE__ */ jsxs(
      "div",
      {
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        ref: forwardedRef,
        className: clsx(
          consumerClassName,
          containerColorsCSS({
            color: derivedColor,
            variant: derivedVariant
          }),
          progressCircleContainerCSS({
            size,
            color: derivedColor,
            variant: derivedVariant
          })
        ),
        style: consumerStyle,
        ...behavioralTrackingProps,
        ...forwardProps,
        "data-dt-component": COMPONENT_NAME,
        children: [
          /* @__PURE__ */ jsx(
            "span",
            {
              id,
              className: clsx(progressCircleRootCSS),
              style: {
                height: `${progressCircleSizes[size]}px`,
                width: `${progressCircleSizes[size]}px`
              },
              ...ariaLabelingProps,
              ...progressAriaProps,
              children: /* @__PURE__ */ jsxs(
                "svg",
                {
                  className: progressCircleSvgCss,
                  viewBox: `0 0 ${progressCircleSizes[size]} ${progressCircleSizes[size]}`,
                  width: "100%",
                  height: "100%",
                  children: [
                    /* @__PURE__ */ jsx(
                      "circle",
                      {
                        fill: "none",
                        className: clsx(progressCircleBackgroundCSS),
                        cx: progressCircleSizes[size] / 2,
                        cy: progressCircleSizes[size] / 2,
                        r: (progressCircleSizes[size] - progressCircleThickness) / 2,
                        strokeWidth: progressCircleThickness
                      }
                    ),
                    /* @__PURE__ */ jsx(
                      "circle",
                      {
                        style: circleStyles,
                        className: progressCircleCSS({ indeterminate: isIndeterminate }),
                        fill: "none",
                        pathLength: isIndeterminate ? 360 : circleStyles["strokeDasharray"],
                        cx: progressCircleSizes[size] / 2,
                        cy: progressCircleSizes[size] / 2,
                        r: (progressCircleSizes[size] - progressCircleThickness) / 2,
                        strokeWidth: progressCircleThickness
                      }
                    )
                  ]
                }
              )
            }
          ),
          children && /* @__PURE__ */ jsx(
            "label",
            {
              className: clsx(
                textStyle({ textStyle: size === "small" ? "small" : "base" })
              ),
              htmlFor: id,
              children
            }
          )
        ]
      }
    );
  }
);
ProgressCircle.displayName = COMPONENT_NAME;
export {
  ProgressCircle
};
