// @dynatrace/strato-components/forms/forms-core/components/form-field-messages/FormFieldMessageBaseItem.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var FormFieldMessageBaseItem_exports = {};
__export(FormFieldMessageBaseItem_exports, {
  InternalBaseItem: () => InternalBaseItem
});
module.exports = __toCommonJS(FormFieldMessageBaseItem_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_strato_icons = require("@dynatrace/strato-icons");
var styles = __toESM(require("./FormFieldMessageItem.sty.js"));
var import_useBehavioralTrackingProps = require("../../../../core/hooks/useBehavioralTrackingProps.js");
var import_Flex = require("../../../../layouts/flex/Flex.js");
var import_useFormFieldMessagesContext = require("../../hooks/useFormFieldMessagesContext.js");
const InternalBaseItem = (props) => {
  const {
    children,
    variant = "hint",
    className: consumerClassName,
    style: consumerStyle,
    disabled: propDisabled = false,
    "data-testid": dataTestId,
    id: propId,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const { disabled: contextDisabled } = (0, import_useFormFieldMessagesContext.useFormFieldMessagesContext)();
  const disabled = propDisabled || contextDisabled;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_Flex.Flex,
    {
      className: (0, import_clsx.default)(
        styles.formFieldMessageItem({ variant, disabled }),
        consumerClassName
      ),
      style: consumerStyle,
      alignItems: "flex-start",
      gap: 4,
      "data-testid": dataTestId,
      id: propId,
      ...behavioralTrackingProps,
      children: [
        variant === "error" && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Flex.Flex, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.CriticalIcon, { size: "small" }) }),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: styles.formFieldMessageItemText, children })
      ]
    }
  );
};
InternalBaseItem.displayName = "FormFieldMessageItem";
