// @dynatrace/strato-components/forms/forms-core/components/form-field-messages/FormFieldMessages.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var FormFieldMessages_exports = {};
__export(FormFieldMessages_exports, {
  FormFieldMessages: () => FormFieldMessages
});
module.exports = __toCommonJS(FormFieldMessages_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_clsx = __toESM(require("clsx"));
var import_react2 = require("react");
var import_FormFieldMessageItem = require("./FormFieldMessageItem.js");
var styles = __toESM(require("./FormFieldMessages.sty.js"));
var import_useBehavioralTrackingProps = require("../../../../core/hooks/useBehavioralTrackingProps.js");
var import_Flex = require("../../../../layouts/flex/Flex.js");
var import_useFormFieldContext = require("../../hooks/useFormFieldContext.js");
var import_useFormFieldMessagesContext = require("../../hooks/useFormFieldMessagesContext.js");
const FormFieldMessagesRoot = /* @__PURE__ */ (0, import_react2.forwardRef)((FormFieldMessageProps, forwardedRef) => {
  const {
    children,
    className: consumerClassName,
    style: consumerStyle,
    "data-dtrum-allow": dataDtrumAllow,
    "data-dtrum-mask": dataDtrumMask,
    "data-testid": dataTestId,
    ...remainingProps
  } = FormFieldMessageProps;
  const [behavioralTrackingProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const { state } = (0, import_useFormFieldMessagesContext.useFormFieldMessagesContext)();
  const { disabled, ariaDisabled } = (0, import_useFormFieldContext.useFormFieldContext)();
  const errorMessages = state.formControlMessages.filter(
    (message) => message.variant === "error"
  );
  const internalHintMessages = state.formControlMessages.filter(
    (message) => message.variant === "hint"
  );
  const externalHintMessages = state.customMessages.filter(
    (message) => message.variant === "hint"
  );
  const hintMessages = [...internalHintMessages, ...externalHintMessages];
  const formMessages = state.showError ? errorMessages : internalHintMessages;
  const hideMessages = !state.showError && hintMessages.length === 0;
  let content = null;
  if (typeof children === "function") {
    content = children(formMessages);
  } else if (children !== void 0 && children !== null) {
    content = children;
  } else {
    content = formMessages.map((message) => /* @__PURE__ */ (0, import_react.createElement)(
      FormFieldMessages.Item,
      {
        disabled: disabled || ariaDisabled,
        ...message,
        key: message.id
      }
    ));
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Flex.Flex,
    {
      ref: forwardedRef,
      flexDirection: "column",
      "aria-live": "polite",
      style: { ...consumerStyle, display: hideMessages ? "none" : "flex" },
      className: (0, import_clsx.default)(consumerClassName, styles.formFieldMessages),
      "aria-hidden": hideMessages,
      "data-dtrum-allow": dataDtrumAllow,
      "data-dtrum-mask": dataDtrumMask,
      "data-testid": dataTestId,
      ...behavioralTrackingProps,
      children: content
    }
  );
});
FormFieldMessagesRoot.displayName = "FormFieldMessages";
const FormFieldMessages = Object.assign(FormFieldMessagesRoot, {
  Item: import_FormFieldMessageItem.FormFieldMessageItem
});
