// @dynatrace/strato-components/tables/DataTable/tanstack-table-core-override/src/core/cell.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var cell_exports = {};
__export(cell_exports, {
  createCell: () => createCell
});
module.exports = __toCommonJS(cell_exports);
var import_utils = require("../utils");
function createCell(table, row, column, columnId) {
  const getRenderValue = () => cell.getValue() ?? table.options.renderFallbackValue;
  const cell = {
    id: `${row.id}_${column.id}`,
    row,
    column,
    getValue: () => row.getValue(columnId),
    renderValue: getRenderValue,
    getContext: (0, import_utils.memo)(
      () => [table, column, row, cell],
      (table2, column2, row2, cell2) => ({
        table: table2,
        column: column2,
        row: row2,
        cell: cell2,
        getValue: cell2.getValue,
        renderValue: cell2.renderValue
      }),
      (0, import_utils.getMemoOptions)(table.options, "debugCells", "cell.getContext")
    )
  };
  table._features.forEach((feature) => {
    feature.createCell?.(
      cell,
      column,
      row,
      table
    );
  }, {});
  return cell;
}
