// @dynatrace/strato-components/tables/DataTable/tanstack-table-core-override/src/core/headers.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var headers_exports = {};
__export(headers_exports, {
  Headers: () => Headers,
  buildHeaderGroups: () => buildHeaderGroups
});
module.exports = __toCommonJS(headers_exports);
var import_utils = require("../utils");
const debug = "debugHeaders";
function createHeader(table, column, options) {
  const id = options.id ?? column.id;
  let header = {
    id,
    column,
    index: options.index,
    isPlaceholder: !!options.isPlaceholder,
    placeholderId: options.placeholderId,
    depth: options.depth,
    subHeaders: [],
    colSpan: 0,
    rowSpan: 0,
    headerGroup: null,
    getLeafHeaders: () => {
      const leafHeaders = [];
      const recurseHeader = (h) => {
        if (h.subHeaders && h.subHeaders.length) {
          h.subHeaders.map(recurseHeader);
        }
        leafHeaders.push(h);
      };
      recurseHeader(header);
      return leafHeaders;
    },
    getContext: () => ({
      table,
      header,
      column
    })
  };
  table._features.forEach((feature) => {
    feature.createHeader?.(header, table);
  });
  return header;
}
const Headers = {
  createTable: (table) => {
    table.getHeaderGroups = (0, import_utils.memo)(
      () => [
        table.getAllColumns(),
        table.getVisibleLeafColumns(),
        table.getState().columnPinning.left,
        table.getState().columnPinning.right
      ],
      (allColumns, leafColumns, left, right) => {
        const leftColumns = left?.map((columnId) => leafColumns.find((d) => d.id === columnId)).filter(Boolean) ?? [];
        const rightColumns = right?.map((columnId) => leafColumns.find((d) => d.id === columnId)).filter(Boolean) ?? [];
        const centerColumns = leafColumns.filter(
          (column) => !left?.includes(column.id) && !right?.includes(column.id)
        );
        const headerGroups = buildHeaderGroups(
          allColumns,
          [...leftColumns, ...centerColumns, ...rightColumns],
          table
        );
        return headerGroups;
      },
      (0, import_utils.getMemoOptions)(table.options, debug, "getHeaderGroups")
    );
    table.getCenterHeaderGroups = (0, import_utils.memo)(
      () => [
        table.getAllColumns(),
        table.getVisibleLeafColumns(),
        table.getState().columnPinning.left,
        table.getState().columnPinning.right
      ],
      (allColumns, leafColumns, left, right) => {
        leafColumns = leafColumns.filter(
          (column) => !left?.includes(column.id) && !right?.includes(column.id)
        );
        return buildHeaderGroups(allColumns, leafColumns, table, "center");
      },
      (0, import_utils.getMemoOptions)(table.options, debug, "getCenterHeaderGroups")
    );
    table.getLeftHeaderGroups = (0, import_utils.memo)(
      () => [
        table.getAllColumns(),
        table.getVisibleLeafColumns(),
        table.getState().columnPinning.left
      ],
      (allColumns, leafColumns, left) => {
        const orderedLeafColumns = left?.map((columnId) => leafColumns.find((d) => d.id === columnId)).filter(Boolean) ?? [];
        return buildHeaderGroups(allColumns, orderedLeafColumns, table, "left");
      },
      (0, import_utils.getMemoOptions)(table.options, debug, "getLeftHeaderGroups")
    );
    table.getRightHeaderGroups = (0, import_utils.memo)(
      () => [
        table.getAllColumns(),
        table.getVisibleLeafColumns(),
        table.getState().columnPinning.right
      ],
      (allColumns, leafColumns, right) => {
        const orderedLeafColumns = right?.map((columnId) => leafColumns.find((d) => d.id === columnId)).filter(Boolean) ?? [];
        return buildHeaderGroups(
          allColumns,
          orderedLeafColumns,
          table,
          "right"
        );
      },
      (0, import_utils.getMemoOptions)(table.options, debug, "getRightHeaderGroups")
    );
    table.getFooterGroups = (0, import_utils.memo)(
      () => [table.getHeaderGroups()],
      (headerGroups) => {
        return [...headerGroups].reverse();
      },
      (0, import_utils.getMemoOptions)(table.options, debug, "getFooterGroups")
    );
    table.getLeftFooterGroups = (0, import_utils.memo)(
      () => [table.getLeftHeaderGroups()],
      (headerGroups) => {
        return [...headerGroups].reverse();
      },
      (0, import_utils.getMemoOptions)(table.options, debug, "getLeftFooterGroups")
    );
    table.getCenterFooterGroups = (0, import_utils.memo)(
      () => [table.getCenterHeaderGroups()],
      (headerGroups) => {
        return [...headerGroups].reverse();
      },
      (0, import_utils.getMemoOptions)(table.options, debug, "getCenterFooterGroups")
    );
    table.getRightFooterGroups = (0, import_utils.memo)(
      () => [table.getRightHeaderGroups()],
      (headerGroups) => {
        return [...headerGroups].reverse();
      },
      (0, import_utils.getMemoOptions)(table.options, debug, "getRightFooterGroups")
    );
    table.getFlatHeaders = (0, import_utils.memo)(
      () => [table.getHeaderGroups()],
      (headerGroups) => {
        return headerGroups.map((headerGroup) => {
          return headerGroup.headers;
        }).flat();
      },
      (0, import_utils.getMemoOptions)(table.options, debug, "getFlatHeaders")
    );
    table.getLeftFlatHeaders = (0, import_utils.memo)(
      () => [table.getLeftHeaderGroups()],
      (left) => {
        return left.map((headerGroup) => {
          return headerGroup.headers;
        }).flat();
      },
      (0, import_utils.getMemoOptions)(table.options, debug, "getLeftFlatHeaders")
    );
    table.getCenterFlatHeaders = (0, import_utils.memo)(
      () => [table.getCenterHeaderGroups()],
      (left) => {
        return left.map((headerGroup) => {
          return headerGroup.headers;
        }).flat();
      },
      (0, import_utils.getMemoOptions)(table.options, debug, "getCenterFlatHeaders")
    );
    table.getRightFlatHeaders = (0, import_utils.memo)(
      () => [table.getRightHeaderGroups()],
      (left) => {
        return left.map((headerGroup) => {
          return headerGroup.headers;
        }).flat();
      },
      (0, import_utils.getMemoOptions)(table.options, debug, "getRightFlatHeaders")
    );
    table.getCenterLeafHeaders = (0, import_utils.memo)(
      () => [table.getCenterFlatHeaders()],
      (flatHeaders) => {
        return flatHeaders.filter((header) => !header.subHeaders?.length);
      },
      (0, import_utils.getMemoOptions)(table.options, debug, "getCenterLeafHeaders")
    );
    table.getLeftLeafHeaders = (0, import_utils.memo)(
      () => [table.getLeftFlatHeaders()],
      (flatHeaders) => {
        return flatHeaders.filter((header) => !header.subHeaders?.length);
      },
      (0, import_utils.getMemoOptions)(table.options, debug, "getLeftLeafHeaders")
    );
    table.getRightLeafHeaders = (0, import_utils.memo)(
      () => [table.getRightFlatHeaders()],
      (flatHeaders) => {
        return flatHeaders.filter((header) => !header.subHeaders?.length);
      },
      (0, import_utils.getMemoOptions)(table.options, debug, "getRightLeafHeaders")
    );
    table.getLeafHeaders = (0, import_utils.memo)(
      () => [
        table.getLeftHeaderGroups(),
        table.getCenterHeaderGroups(),
        table.getRightHeaderGroups()
      ],
      (left, center, right) => {
        return [
          ...left[0]?.headers ?? [],
          ...center[0]?.headers ?? [],
          ...right[0]?.headers ?? []
        ].map((header) => {
          return header.getLeafHeaders();
        }).flat();
      },
      (0, import_utils.getMemoOptions)(table.options, debug, "getLeafHeaders")
    );
  }
};
function buildHeaderGroups(allColumns, columnsToGroup, table, headerFamily) {
  let maxDepth = 0;
  const findMaxDepth = (columns, depth = 1) => {
    maxDepth = Math.max(maxDepth, depth);
    columns.filter((column) => column.getIsVisible()).forEach((column) => {
      if (column.columns?.length) {
        findMaxDepth(column.columns, depth + 1);
      }
    }, 0);
  };
  findMaxDepth(allColumns);
  let headerGroups = [];
  const createHeaderGroup = (headersToGroup, depth) => {
    const headerGroup = {
      depth,
      id: [headerFamily, `${depth}`].filter(Boolean).join("_"),
      headers: []
    };
    const pendingParentHeaders = [];
    headersToGroup.forEach((headerToGroup) => {
      const latestPendingParentHeader = [...pendingParentHeaders].reverse()[0];
      const isLeafHeader = headerToGroup.column.depth === headerGroup.depth;
      let column;
      let isPlaceholder = false;
      if (isLeafHeader && headerToGroup.column.parent) {
        column = headerToGroup.column.parent;
      } else {
        column = headerToGroup.column;
        isPlaceholder = true;
      }
      if (latestPendingParentHeader && latestPendingParentHeader?.column === column) {
        latestPendingParentHeader.subHeaders.push(headerToGroup);
      } else {
        const header = createHeader(table, column, {
          id: [headerFamily, depth, column.id, headerToGroup?.id].filter(Boolean).join("_"),
          isPlaceholder,
          placeholderId: isPlaceholder ? `${pendingParentHeaders.filter((d) => d.column === column).length}` : void 0,
          depth,
          index: pendingParentHeaders.length
        });
        header.subHeaders.push(headerToGroup);
        pendingParentHeaders.push(header);
      }
      headerGroup.headers.push(headerToGroup);
      headerToGroup.headerGroup = headerGroup;
    });
    headerGroups.push(headerGroup);
    if (depth > 0) {
      createHeaderGroup(pendingParentHeaders, depth - 1);
    }
  };
  const bottomHeaders = columnsToGroup.map(
    (column, index) => createHeader(table, column, {
      depth: maxDepth,
      index
    })
  );
  createHeaderGroup(bottomHeaders, maxDepth - 1);
  headerGroups.reverse();
  const recurseHeadersForSpans = (headers) => {
    const filteredHeaders = headers.filter(
      (header) => header.column.getIsVisible()
    );
    return filteredHeaders.map((header) => {
      let colSpan = 0;
      let rowSpan = 0;
      let childRowSpans = [0];
      if (header.subHeaders && header.subHeaders.length) {
        childRowSpans = [];
        recurseHeadersForSpans(header.subHeaders).forEach(
          ({ colSpan: childColSpan, rowSpan: childRowSpan }) => {
            colSpan += childColSpan;
            childRowSpans.push(childRowSpan);
          }
        );
      } else {
        colSpan = 1;
      }
      const minChildRowSpan = Math.min(...childRowSpans);
      rowSpan = rowSpan + minChildRowSpan;
      header.colSpan = colSpan;
      header.rowSpan = rowSpan;
      return { colSpan, rowSpan };
    });
  };
  recurseHeadersForSpans(headerGroups[0]?.headers ?? []);
  return headerGroups;
}
