// @dynatrace/strato-components/tables/DataTable/tanstack-table-core-override/src/core/table.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var table_exports = {};
__export(table_exports, {
  createTable: () => createTable
});
module.exports = __toCommonJS(table_exports);
var import_utils = require("../utils");
var import_column = require("./column");
var import_headers = require("./headers");
var import_ColumnFaceting = require("../features/ColumnFaceting");
var import_ColumnFiltering = require("../features/ColumnFiltering");
var import_ColumnGrouping = require("../features/ColumnGrouping");
var import_ColumnOrdering = require("../features/ColumnOrdering");
var import_ColumnPinning = require("../features/ColumnPinning");
var import_ColumnSizing = require("../features/ColumnSizing");
var import_ColumnVisibility = require("../features/ColumnVisibility");
var import_GlobalFaceting = require("../features/GlobalFaceting");
var import_GlobalFiltering = require("../features/GlobalFiltering");
var import_RowExpanding = require("../features/RowExpanding");
var import_RowPagination = require("../features/RowPagination");
var import_RowPinning = require("../features/RowPinning");
var import_RowSelection = require("../features/RowSelection");
var import_RowSorting = require("../features/RowSorting");
const builtInFeatures = [
  import_headers.Headers,
  import_ColumnVisibility.ColumnVisibility,
  import_ColumnOrdering.ColumnOrdering,
  import_ColumnPinning.ColumnPinning,
  import_ColumnFaceting.ColumnFaceting,
  import_ColumnFiltering.ColumnFiltering,
  import_GlobalFaceting.GlobalFaceting,
  //depends on ColumnFaceting
  import_GlobalFiltering.GlobalFiltering,
  //depends on ColumnFiltering
  import_RowSorting.RowSorting,
  import_ColumnGrouping.ColumnGrouping,
  //depends on RowSorting
  import_RowExpanding.RowExpanding,
  import_RowPagination.RowPagination,
  import_RowPinning.RowPinning,
  import_RowSelection.RowSelection,
  import_ColumnSizing.ColumnSizing
];
function createTable(options) {
  if (options.debugAll || options.debugTable) {
    console.info("Creating Table Instance...");
  }
  const _features = [...builtInFeatures, ...options._features ?? []];
  let table = { _features };
  const defaultOptions = table._features.reduce((obj, feature) => {
    return Object.assign(obj, feature.getDefaultOptions?.(table));
  }, {});
  const mergeOptions = (options2) => {
    if (table.options.mergeOptions) {
      return table.options.mergeOptions(defaultOptions, options2);
    }
    return {
      ...defaultOptions,
      ...options2
    };
  };
  const coreInitialState = {};
  let initialState = {
    ...coreInitialState,
    ...options.initialState ?? {}
  };
  table._features.forEach((feature) => {
    initialState = feature.getInitialState?.(initialState) ?? initialState;
  });
  const queued = [];
  let queuedTimeout = false;
  const coreInstance = {
    _features,
    options: {
      ...defaultOptions,
      ...options
    },
    initialState,
    _queue: (cb) => {
      queued.push(cb);
      if (!queuedTimeout) {
        queuedTimeout = true;
        Promise.resolve().then(() => {
          while (queued.length) {
            queued.shift()();
          }
          queuedTimeout = false;
        }).catch(
          (error) => setTimeout(() => {
            throw error;
          })
        );
      }
    },
    reset: () => {
      table.setState(table.initialState);
    },
    setOptions: (updater) => {
      const newOptions = (0, import_utils.functionalUpdate)(updater, table.options);
      table.options = mergeOptions(newOptions);
    },
    getState: () => {
      return table.options.state;
    },
    setState: (updater) => {
      table.options.onStateChange?.(updater);
    },
    _getRowId: (row, index, parent) => table.options.getRowId?.(row, index, parent) ?? `${parent ? [parent.id, index].join(".") : index}`,
    getCoreRowModel: () => {
      if (!table._getCoreRowModel) {
        table._getCoreRowModel = table.options.getCoreRowModel(table);
      }
      return table._getCoreRowModel();
    },
    // The final calls start at the bottom of the model,
    // expanded rows, which then work their way up
    getRowModel: () => {
      return table.getPaginationRowModel();
    },
    //in next version, we should just pass in the row model as the optional 2nd arg
    getRow: (id, searchAll) => {
      let row = (searchAll ? table.getPrePaginationRowModel() : table.getRowModel()).rowsById[id];
      if (!row) {
        row = table.getCoreRowModel().rowsById[id];
        if (!row) {
          if (true) {
            throw new Error(`getRow could not find row with ID: ${id}`);
          }
          throw new Error();
        }
      }
      return row;
    },
    _getDefaultColumnDef: (0, import_utils.memo)(
      () => [table.options.defaultColumn],
      (defaultColumn) => {
        defaultColumn = defaultColumn ?? {};
        return {
          header: (props) => {
            const resolvedColumnDef = props.header.column.columnDef;
            if (resolvedColumnDef.accessorKey) {
              return resolvedColumnDef.accessorKey;
            }
            if (resolvedColumnDef.accessorFn) {
              return resolvedColumnDef.id;
            }
            return null;
          },
          // footer: props => props.header.column.id,
          cell: (props) => props.renderValue()?.toString?.() ?? null,
          ...table._features.reduce((obj, feature) => {
            return Object.assign(obj, feature.getDefaultColumnDef?.());
          }, {}),
          ...defaultColumn
        };
      },
      (0, import_utils.getMemoOptions)(options, "debugColumns", "_getDefaultColumnDef")
    ),
    _getColumnDefs: () => table.options.columns,
    getAllColumns: (0, import_utils.memo)(
      () => [table._getColumnDefs()],
      (columnDefs) => {
        const recurseColumns = (columnDefs2, parent, depth = 0) => {
          return columnDefs2.map((columnDef) => {
            const column = (0, import_column.createColumn)(table, columnDef, depth, parent);
            const groupingColumnDef = columnDef;
            column.columns = groupingColumnDef.columns ? recurseColumns(groupingColumnDef.columns, column, depth + 1) : [];
            return column;
          });
        };
        return recurseColumns(columnDefs);
      },
      (0, import_utils.getMemoOptions)(options, "debugColumns", "getAllColumns")
    ),
    getAllFlatColumns: (0, import_utils.memo)(
      () => [table.getAllColumns()],
      (allColumns) => {
        return allColumns.flatMap((column) => {
          return column.getFlatColumns();
        });
      },
      (0, import_utils.getMemoOptions)(options, "debugColumns", "getAllFlatColumns")
    ),
    _getAllFlatColumnsById: (0, import_utils.memo)(
      () => [table.getAllFlatColumns()],
      (flatColumns) => {
        return flatColumns.reduce((acc, column) => {
          acc[column.id] = column;
          return acc;
        }, {});
      },
      (0, import_utils.getMemoOptions)(options, "debugColumns", "getAllFlatColumnsById")
    ),
    getAllLeafColumns: (0, import_utils.memo)(
      () => [table.getAllColumns(), table._getOrderColumnsFn()],
      (allColumns, orderColumns) => {
        let leafColumns = allColumns.flatMap(
          (column) => column.getLeafColumns()
        );
        return orderColumns(leafColumns);
      },
      (0, import_utils.getMemoOptions)(options, "debugColumns", "getAllLeafColumns")
    ),
    getColumn: (columnId) => {
      const column = table._getAllFlatColumnsById()[columnId];
      if (!column) {
        console.error(`[Table] Column with id '${columnId}' does not exist.`);
      }
      return column;
    }
  };
  Object.assign(table, coreInstance);
  for (let index = 0; index < table._features.length; index++) {
    const feature = table._features[index];
    feature?.createTable?.(table);
  }
  return table;
}
