// @dynatrace/strato-components/esm/typography/index.js
import { Blockquote } from "./block-quote/Blockquote.js";
import { Code } from "./code/Code.js";
import { Emphasis } from "./emphasis/Emphasis.js";
import { ExternalLink } from "./external-link/ExternalLink.js";
import { Heading } from "./heading/Heading.js";
import { Highlight } from "./highlight/Highlight.js";
import { Link } from "./link/Link.js";
import { List } from "./list/List.js";
import { Paragraph } from "./paragraph/Paragraph.js";
import { Strikethrough } from "./strikethrough/Strikethrough.js";
import { Strong } from "./strong/Strong.js";
import { Text } from "./text/Text.js";
import { TextEllipsis } from "./text-ellipsis/TextEllipsis.js";
import {
  calculateEllipsisInfo,
  getTextHeight,
  centerEllipsizeText,
  getTextWidth,
  ELLIPSIS_APPROXIMATION_THRESHOLD,
  ELLIPSIS_CHARACTER
} from "./utils.js";
export {
  Blockquote,
  Code,
  Emphasis,
  ExternalLink,
  Heading,
  Highlight,
  Link,
  List,
  Paragraph,
  Strikethrough,
  Strong,
  Text,
  TextEllipsis,
  ELLIPSIS_APPROXIMATION_THRESHOLD as _ELLIPSIS_APPROXIMATION_THRESHOLD,
  ELLIPSIS_CHARACTER as _ELLIPSIS_CHARACTER,
  calculateEllipsisInfo as _calculateEllipsisInfo,
  centerEllipsizeText as _centerEllipsizeText,
  getTextHeight as _getTextHeight,
  getTextWidth as _getTextWidth
};
