// @dynatrace/strato-components/charts/categorical-bar/displays/BarChartDisplay.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var BarChartDisplay_exports = {};
__export(BarChartDisplay_exports, {
  BarChartDisplay: () => BarChartDisplay
});
module.exports = __toCommonJS(BarChartDisplay_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_bar_style = require("./bar-style.js");
var import_BarCanvas = require("../../core/components/canvas/components/BarCanvas.js");
var import_useGridPlotSize = require("../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_safe_get = require("../../core/utils/safe-get.js");
var import_useCategoricalBarChartBaseConfig = require("../hooks/useCategoricalBarChartBaseConfig.js");
var import_useCategoricalBarFilteredData = require("../hooks/useCategoricalBarFilteredData.js");
var import_useDisplayConfigContext = require("../hooks/useDisplayConfigContext.js");
var import_categorical_layout = require("../utils/categorical-layout.js");
var import_layout = require("../utils/layout.js");
const BarChartDisplay = () => {
  const { highlightedDimension, categoryScale, dimensionScale, valueScale } = (0, import_useDisplayConfigContext.useDisplayConfigContext)();
  const { filteredCategories, isSingleDimension, dimensionMap, categoryMap } = (0, import_useCategoricalBarFilteredData.useCategoricalBarFilteredData)();
  const { layout, groupMode } = (0, import_useCategoricalBarChartBaseConfig.useCategoricalBarChartBaseConfig)();
  const { width, height } = (0, import_useGridPlotSize.useGridPlotSize)();
  const vertical = (0, import_layout.isVertical)(layout);
  const { categoryCoord, valueCoord, barSize, barThickness } = (0, import_categorical_layout.getBarCoordinates)(
    groupMode,
    categoryScale,
    valueScale,
    dimensionScale,
    isSingleDimension,
    dimensionMap,
    categoryMap
  );
  const widthAccessor = vertical ? barThickness : barSize;
  const heightAccessor = vertical ? barSize : barThickness;
  const color = (dimension) => {
    const { name, color: color2 } = (0, import_safe_get.safeGet)(dimensionMap, dimension);
    return (0, import_bar_style.getBarStyle)(name, color2, highlightedDimension).color;
  };
  const opacity = (dimension) => {
    const { name, color: color2 } = (0, import_safe_get.safeGet)(dimensionMap, dimension);
    return (0, import_bar_style.getBarStyle)(name, color2, highlightedDimension).opacity;
  };
  return filteredCategories.map((category) => {
    const y = vertical ? valueCoord : (d) => categoryCoord(d, category);
    const x = vertical ? (d) => categoryCoord(d, category) : valueCoord;
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_BarCanvas.BarCanvas,
      {
        data: category.dimensions,
        canvasWidth: width,
        canvasHeight: height,
        x,
        y,
        width: widthAccessor,
        height: heightAccessor,
        isVertical: vertical,
        strokeWidth: 1,
        color,
        opacity
      },
      category.id
    );
  });
};
BarChartDisplay["displayName"] = "BarChartDisplay";
