// @dynatrace/strato-components/esm/charts/pie/components/PieChartOverlays/PieChartOverlays.js
import { jsx, jsxs } from "react/jsx-runtime";
import { container } from "./PieChartOverlays.sty.js";
import { Flex } from "../../../../layouts/flex/Flex.js";
import { useChartLayoutLoading } from "../../../core/components/chart-layout/hooks/useChartLayoutLoading.js";
import { useOverlay } from "../../../core/components/overlay-tooltip/contexts/OverlayContext.js";
import { isToolbarEnabled } from "../../../core/utils/is-toolbar-enabled.js";
import { useToolbarConfig } from "../../context/ToolbarContext.js";
import { useOverlayHandlers } from "../../hooks/useOverlayHandlers.js";
import { PieChartToolbar } from "../PieChartToolbar/PieChartToolbar.js";
import { PieChartTooltip } from "../PieChartTooltip/PieChartTooltip.js";
const PieChartOverlays = ({ children }) => {
  const overlay = useOverlay();
  const loading = useChartLayoutLoading();
  const { onMouseEnterHandler, onMouseLeaveHandler } = useOverlayHandlers();
  const toolbarConfig = useToolbarConfig();
  const toolbarEnabled = isToolbarEnabled(toolbarConfig);
  return /* @__PURE__ */ jsxs(
    Flex,
    {
      ref: overlay.overlayRef,
      "data-testid": "overlays-wrapper",
      justifyContent: "center",
      alignItems: "center",
      className: container,
      onMouseEnter: loading ? void 0 : onMouseEnterHandler,
      onMouseLeave: loading ? void 0 : onMouseLeaveHandler,
      children: [
        !loading && /* @__PURE__ */ jsx(PieChartTooltip, {}),
        !loading && toolbarEnabled && /* @__PURE__ */ jsx(PieChartToolbar, {}),
        children
      ]
    }
  );
};
export {
  PieChartOverlays
};
