// @dynatrace/strato-components/esm/charts/sparkline/Sparkline.js
import { jsx } from "react/jsx-runtime";
import { ThresholdIndicator } from "../core/components/sparkline-display/slots/ThresholdIndicator.js";
import { XAxis } from "../core/components/sparkline-display/slots/XAxis.js";
import { YAxis } from "../core/components/sparkline-display/slots/YAxis.js";
import { SparklineDisplay } from "../core/components/sparkline-display/SparklineDisplay.js";
import { iterateSparklineSlots } from "../core/components/sparkline-display/utils/iterate-sparkline-slots.js";
import { EmptyStateSlot } from "../core/slots/chart-states/EmptyStateSlot.js";
import { ErrorStateSlot } from "../core/slots/chart-states/ErrorStateSlot.js";
const SparkLineRoot = ({
  children,
  ...props
}) => {
  const slots = iterateSparklineSlots(children);
  return /* @__PURE__ */ jsx(
    SparklineDisplay,
    {
      ...props,
      thresholds: slots.thresholds,
      xAxis: slots.xAxis,
      yAxis: slots.yAxis,
      children
    }
  );
};
SparkLineRoot["displayName"] = "Sparkline";
const Sparkline = Object.assign(SparkLineRoot, {
  ThresholdIndicator,
  EmptyState: EmptyStateSlot,
  ErrorState: ErrorStateSlot,
  XAxis,
  YAxis
});
export {
  Sparkline
};
