// @dynatrace/strato-components/content/keyboard-shortcut-tooltip/KeyboardShortcutTooltip.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var KeyboardShortcutTooltip_exports = {};
__export(KeyboardShortcutTooltip_exports, {
  KeyboardShortcutTooltip: () => KeyboardShortcutTooltip
});
module.exports = __toCommonJS(KeyboardShortcutTooltip_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_typography = __toESM(require("@dynatrace/strato-design-tokens/typography"));
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useSafeForwardProps = require("../../core/hooks/useSafeForwardProps.js");
var import_Flex = require("../../layouts/flex/Flex.js");
var import_Tooltip = require("../../overlays/tooltip/Tooltip.js");
var import_Text = require("../../typography/text/Text.js");
var import_KeyboardShortcut = require("../keyboard-shortcut/KeyboardShortcut.js");
const COMPONENT_NAME = "KeyboardShortcutTooltip";
const KeyboardShortcutTooltip = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    children,
    keys,
    shortcutDisabled,
    text,
    placement = "top",
    open,
    defaultOpen,
    onOpenChange,
    delay = "default",
    disabled,
    style: consumerStyle,
    className: consumerClassName,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const ariaLabelingProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(restProps);
  const forwardProps = (0, import_useSafeForwardProps.useSafeForwardProps)(restProps, ariaLabelingProps);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Tooltip.Tooltip,
    {
      ref: forwardedRef,
      placement,
      open,
      defaultOpen,
      onOpenChange,
      delay,
      disabled,
      text: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { flexDirection: "column", alignItems: "center", gap: 4, children: [
        text && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_Text.Text,
          {
            textStyle: "small-emphasized",
            style: { maxWidth: "100%", overflowWrap: "anywhere" },
            children: text
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_KeyboardShortcut.KeyboardShortcut,
          {
            keys,
            variant: "minimal",
            disabled: shortcutDisabled,
            style: {
              lineHeight: import_typography.default.Text.Small.Default.LineHeight
            }
          }
        )
      ] }),
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...ariaLabelingProps,
      ...behavioralTrackingProps,
      ...forwardProps,
      "data-dt-component": COMPONENT_NAME,
      children
    }
  );
});
KeyboardShortcutTooltip.displayName = COMPONENT_NAME;
