// @dynatrace/strato-geo/esm/map/components/ConnectionLayer/ConnectionLayer.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { _useBehavioralTrackingProps as useBehavioralTrackingProps } from "@dynatrace/strato-components/core";
import { ConnectionLayerDirection } from "./ConnectionLayerDirection.js";
import { ConnectionLayerLine } from "./ConnectionLayerLine.js";
import { ConnectionLayerTooltip } from "./ConnectionLayerTooltip.js";
import { parseConnectionDataToGeoJSON } from "./utils/parse-connection-data-to-geo-json.js";
import { DEFAULT_LINE_COLOR, DEFAULT_LINE_THICKNESS } from "../../constants.js";
import { useLayerColoringStrategy } from "../../hooks/use-layer-coloring-strategy.js";
import { useLegendInteractionHighlighting } from "../../hooks/use-legend-interaction-highlighting.js";
import { useLegendInteractionVisibility } from "../../hooks/use-legend-interaction-visibility.js";
import { useResolveConnectionColor } from "../../hooks/use-resolve-color.js";
import { useTooltipTemplate } from "../../hooks/use-tooltip-template.js";
import { ConnectionLayerTooltip as ConnectionLayerTooltipSlot } from "../../slots/ConnectionLayerTooltip.js";
import { isConnectionTooltipTemplate } from "../../utils/tooltip-type-guards.js";
import { TooltipWrapper } from "../TooltipWrapper.js";
const ConnectionLayer = (props) => {
  const [behavioralTrackingProps] = useBehavioralTrackingProps(props);
  const {
    data,
    layerId,
    children,
    curve,
    directionIndicator,
    line,
    thickness = DEFAULT_LINE_THICKNESS
  } = props;
  const colorParser = useLayerColoringStrategy();
  const connectionColor = useResolveConnectionColor(
    DEFAULT_LINE_COLOR,
    props,
    colorParser
  );
  const sourceId = `source-${layerId}`;
  useLegendInteractionHighlighting(props, sourceId);
  useLegendInteractionVisibility(props, layerId);
  const geoJsonData = parseConnectionDataToGeoJSON(
    data,
    thickness,
    connectionColor,
    curve
  );
  const {
    handler: tooltipTemplate,
    hidden: tooltipHidden,
    symbolAlignment
  } = useTooltipTemplate(children, ConnectionLayerTooltipSlot, {
    extractProps: true
  });
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(
      ConnectionLayerLine,
      {
        geoJsonData,
        layerId,
        sourceId,
        line,
        ...behavioralTrackingProps,
        children: /* @__PURE__ */ jsx(TooltipWrapper, { layerId, children: /* @__PURE__ */ jsx(
          ConnectionLayerTooltip,
          {
            tooltipTemplate: isConnectionTooltipTemplate(tooltipTemplate) ? tooltipTemplate : void 0,
            hidden: tooltipHidden,
            symbolAlignment
          }
        ) })
      }
    ),
    directionIndicator && /* @__PURE__ */ jsx(
      ConnectionLayerDirection,
      {
        geoJsonData,
        layerId,
        directionIndicator
      }
    )
  ] });
};
ConnectionLayer["displayName"] = "ConnectionLayer";
export {
  ConnectionLayer
};
