// @dynatrace/strato-geo/esm/map/components/ConnectionLayer/ConnectionLayerLine.js
import { jsx, jsxs } from "react/jsx-runtime";
import { Layer, Source } from "@vis.gl/react-maplibre";
import { _useBehavioralTrackingProps as useBehavioralTrackingProps } from "@dynatrace/strato-components/core";
import {
  DEFAULT_LINE_CAP,
  DEFAULT_LINE_DASH_ARRAY,
  DEFAULT_LINE_JOIN,
  SHAPE_OPACITY_DEFAULT,
  SHAPE_OPACITY_DIMMED
} from "../../constants.js";
const ConnectionLayerLine = (props) => {
  const { geoJsonData, layerId, sourceId, line, children, ...restProps } = props;
  const [behavioralTrackingProps] = useBehavioralTrackingProps(restProps);
  return /* @__PURE__ */ jsxs(Source, { id: sourceId, type: "geojson", data: geoJsonData, promoteId: "id", children: [
    /* @__PURE__ */ jsx(
      Layer,
      {
        type: "line",
        id: layerId,
        layout: {
          "line-cap": DEFAULT_LINE_CAP,
          "line-join": DEFAULT_LINE_JOIN
        },
        paint: {
          "line-color": [
            "case",
            ["boolean", ["feature-state", "hover"], false],
            ["get", "__lineHoveredColor"],
            ["get", "__lineColor"]
          ],
          "line-opacity": [
            "case",
            ["boolean", ["feature-state", "legendHover"], true],
            SHAPE_OPACITY_DEFAULT,
            SHAPE_OPACITY_DIMMED
          ],
          "line-width": ["get", "__lineWidth"],
          ...line && line === "dashed" ? { "line-dasharray": DEFAULT_LINE_DASH_ARRAY } : null
        },
        ...behavioralTrackingProps
      }
    ),
    children
  ] });
};
ConnectionLayerLine["displayName"] = "ConnectionLayerLine";
export {
  ConnectionLayerLine
};
