// @dynatrace/strato-geo/esm/map/components/ConnectionLayer/ConnectionLayerDirection.js
import { jsx } from "react/jsx-runtime";
import { Layer, Source } from "@vis.gl/react-maplibre";
import { SorterIcon } from "@dynatrace/strato-icons";
import { parseDirectionDataToGeoJSON } from "./utils/parse-direction-data-to-geo-json.js";
import {
  DIRECTION_ICON_OUTPUT_SIZE,
  DIRECTION_ICON_SHRINK_RATIO,
  SHAPE_OPACITY_DEFAULT,
  SHAPE_OPACITY_DIMMED
} from "../../constants.js";
import { useLoadSymbolIntoMap } from "../../hooks/use-load-symbol-into-map.js";
const ConnectionLayerDirection = (props) => {
  const { geoJsonData, layerId: connectionLayerId, directionIndicator } = props;
  const layerId = `${connectionLayerId}-direction`;
  const sourceId = `source-${layerId}`;
  const { symbolName } = useLoadSymbolIntoMap(
    /* @__PURE__ */ jsx(SorterIcon, {}),
    layerId,
    DIRECTION_ICON_OUTPUT_SIZE
  );
  const parsedGeoJsonData = parseDirectionDataToGeoJSON({
    geoJsonData,
    directionIndicator
  });
  const iconRotationDegrees = 180;
  return /* @__PURE__ */ jsx(
    Source,
    {
      id: sourceId,
      type: "geojson",
      data: parsedGeoJsonData,
      promoteId: "id",
      children: /* @__PURE__ */ jsx(
        Layer,
        {
          id: layerId,
          beforeId: connectionLayerId,
          type: "symbol",
          layout: {
            "symbol-placement": "point",
            "icon-image": symbolName,
            "icon-rotate": ["+", ["get", "__angle"], iconRotationDegrees],
            "icon-rotation-alignment": "map",
            "icon-allow-overlap": true,
            "icon-ignore-placement": true,
            "icon-size": [
              "*",
              ["get", "__lineWidth"],
              DIRECTION_ICON_SHRINK_RATIO
            ]
          },
          paint: {
            "icon-color": [
              "case",
              ["boolean", ["feature-state", "hover"], false],
              ["get", "__lineHoveredColor"],
              ["get", "__lineColor"]
            ],
            "icon-opacity": [
              "case",
              ["boolean", ["feature-state", "legendHover"], true],
              SHAPE_OPACITY_DEFAULT,
              SHAPE_OPACITY_DIMMED
            ]
          }
        }
      )
    }
  );
};
ConnectionLayerDirection["displayName"] = "ConnectionLayerDirection";
export {
  ConnectionLayerDirection
};
