// @dynatrace/strato-components/esm/charts/single-value-grid/components/SingleValueDisplay.js
import { jsx } from "react/jsx-runtime";
import { isUndefined } from "lodash-es";
import { useIntl } from "react-intl";
import { SingleValueRenderer } from "../../core/components/single-value-renderer/SingleValueRenderer.js";
import { SingleValueGridMessages } from "../messages.js";
import { buildConfig } from "../utils/build-config.js";
import { buildPropsFromGrid } from "../utils/build-props-from-grid.js";
import { getConfigFromAccessors } from "../utils/get-config-from-accessors.js";
import { getDataFromAccessors } from "../utils/get-data-from-accessors.js";
const SingleValueDisplay = ({
  record,
  propsConfig,
  accessors
}) => {
  const intl = useIntl();
  const gridIntents = propsConfig.intents.grid;
  const singleIntents = propsConfig.intents.single;
  const propsFromGrid = buildPropsFromGrid(
    propsConfig.toolbar,
    singleIntents,
    gridIntents,
    intl.formatMessage(SingleValueGridMessages.gridIntentsLabel)
  );
  if (typeof record === "string" || typeof record === "number") {
    const config2 = buildConfig(propsConfig);
    return /* @__PURE__ */ jsx(
      SingleValueRenderer,
      {
        data: record,
        color: config2.color,
        formatter: config2.value?.formatter,
        unit: config2.value?.unit,
        colorRules: config2.colorRules,
        alignment: config2.alignment,
        trendProps: config2.trend,
        applyThresholdBackground: config2.applyThresholdBackground,
        propsFromGrid: propsFromGrid({ value: record })
      }
    );
  }
  const { label, prefixIcon, value, trendData, sparkline } = getDataFromAccessors(record, accessors);
  const accessorsConfig = getConfigFromAccessors(record, accessors);
  const config = buildConfig(propsConfig, accessorsConfig);
  let sparklineProps = void 0;
  if (!isUndefined(sparkline?.data)) {
    sparklineProps = {
      ...config.sparkline,
      data: sparkline.data,
      xAxis: sparkline.xAxis,
      yAxis: sparkline.yAxis
    };
  }
  let trendProps = void 0;
  if (!isUndefined(config.trend) || !isUndefined(trendData)) {
    trendProps = {
      ...config.trend,
      value: trendData?.value,
      label: trendData?.label,
      direction: trendData?.direction
    };
  }
  return /* @__PURE__ */ jsx(
    SingleValueRenderer,
    {
      data: value,
      dataItem: record,
      label,
      prefixIcon,
      color: config.color,
      formatter: config.value?.formatter,
      unit: config.value?.unit,
      colorRules: config.colorRules,
      alignment: config.alignment,
      sparklineProps,
      trendProps,
      applyThresholdBackground: config.applyThresholdBackground,
      propsFromGrid: propsFromGrid(record)
    }
  );
};
export {
  SingleValueDisplay
};
