// @dynatrace/strato-components/charts/timeseries/components/overlays/TimeseriesChartOverlays.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TimeseriesChartOverlays_exports = {};
__export(TimeseriesChartOverlays_exports, {
  TimeseriesChartOverlays: () => TimeseriesChartOverlays
});
module.exports = __toCommonJS(TimeseriesChartOverlays_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_useApplyPanDelta = require("./hooks/useApplyPanDelta.js");
var import_useApplyZoomDelta = require("./hooks/useApplyZoomDelta.js");
var import_useResetInteractions = require("./hooks/useResetInteractions.js");
var import_useResetOnResize = require("./hooks/useResetOnResize.js");
var import_TimeseriesCrosshair = require("./TimeseriesExplore/TimeseriesCrosshair.js");
var import_TimeseriesInspect = require("./TimeseriesExplore/TimeseriesInspect.js");
var import_TimeseriesSelectionArea = require("./TimeseriesSelection/TimeseriesSelectionArea.js");
var import_get_points_parsed = require("./utils/get-points-parsed.js");
var import_constants = require("./utils/tooltip/constants.js");
var import_useGridPlotSize = require("../../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_InteractionsOverlay = require("../../../core/components/interactions-overlay/InteractionsOverlay.js");
var import_types = require("../../../core/components/interactions-overlay/types.js");
var import_is_own_event = require("../../../core/components/interactions-overlay/utils/is-own-event.js");
var import_explore_reducers = require("../../../core/components/interactions-state-helpers/reducers/explore-reducers.js");
var import_mode_reducers = require("../../../core/components/interactions-state-helpers/reducers/mode-reducers.js");
var import_SynchronizedSelectionArea = require("../../../core/components/shared-chart-interactions/components/SynchronizedSelectionArea.js");
var import_VerticalSynchronizedCrosshair = require("../../../core/components/shared-chart-interactions/components/VerticalSynchronizedCrosshair.js");
var import_useSynchronizedCrosshair = require("../../../core/components/shared-chart-interactions/hooks/useSynchronizedCrosshair.js");
var import_useSynchronizedSelection = require("../../../core/components/shared-chart-interactions/hooks/useSynchronizedSelection.js");
var import_useToolbarHandlerRef = require("../../../core/components/toolbar/hooks/useToolbarHandlerRef.js");
var import_get_percentage_from_pixel = require("../../../core/utils/get-percentage-from-pixel.js");
var import_constants2 = require("../../constants.js");
var import_TimeseriesChartOverlayRefContext = require("../../context/TimeseriesChartOverlayRefContext.js");
var import_TimeseriesChartSelectionAreaRef_context = require("../../context/TimeseriesChartSelectionAreaRef.context.js");
var import_selectors = require("../../hooks/store/selectors.js");
var import_state_setters = require("../../hooks/store/state-setters.js");
var import_useChartInteractions = require("../../hooks/useChartInteractions.js");
var import_useChartToolbar = require("../../hooks/useChartToolbar.js");
var import_useTimeseriesChartId = require("../../hooks/useTimeseriesChartId.js");
var import_useTimeseriesFilteredData = require("../../hooks/useTimeseriesFilteredData.js");
var import_useXScale = require("../../hooks/useXScale.js");
var import_useYScales = require("../../hooks/useYScales.js");
var import_TimeseriesChartToolbar = require("../toolbar/TimeseriesChartToolbar.js");
const TimeseriesChartOverlays = () => {
  const handlerRef = (0, import_useToolbarHandlerRef.useToolbarHandlerRef)();
  const { zoomX, pan, zoom } = (0, import_useChartInteractions.useChartInteractions)();
  const selectionRef = (0, import_react.useRef)(null);
  const { width, height } = (0, import_useGridPlotSize.useGridPlotSize)();
  const toolbar = (0, import_useChartToolbar.useChartToolbar)();
  const exploreState = (0, import_selectors.useExploreState)();
  const selectState = (0, import_selectors.useSelectState)();
  const mode = (0, import_selectors.useInteractionModeState)();
  const zoomState = (0, import_selectors.useZoomState)();
  const setState = (0, import_selectors.useSetState)();
  const applyPanDelta = (0, import_useApplyPanDelta.useApplyPanDelta)();
  const applyZoomDelta = (0, import_useApplyZoomDelta.useApplyZoomDelta)();
  const resetInteractions = (0, import_useResetInteractions.useResetInteractions)();
  (0, import_useResetOnResize.useResetOnResize)();
  const synchronizedCrosshair = (0, import_useSynchronizedCrosshair.useSynchronizedCrosshair)();
  const synchronizedSelection = (0, import_useSynchronizedSelection.useSynchronizedSelection)();
  const chartId = (0, import_useTimeseriesChartId.useTimeseriesChartId)();
  const overlayRef = (0, import_react.useRef)(null);
  const {
    updateHoverState: updateSynchronizedCrosshair,
    hoveredChartId,
    disabled: synchronizedCrosshairDisabled,
    setDisabled: disableSynchronizedCrosshair
  } = synchronizedCrosshair ?? {};
  const {
    updateSelectionState: updateSynchronizedSelection,
    resetSelectionState: resetSynchronizedSelection,
    selectedChartId,
    disabled: synchronizedSelectionDisabled
  } = "selectedChartId" in synchronizedSelection ? synchronizedSelection : { disabled: true };
  (0, import_react.useEffect)(() => {
    if ((0, import_lodash_es.every)([
      selectState,
      !synchronizedSelection.selectedChartId,
      !synchronizedSelectionDisabled
    ])) {
      setState((s) => ({ ...s, select: void 0 }));
    }
  }, [synchronizedSelection.selectedChartId]);
  (0, import_react.useEffect)(() => {
    if (synchronizedSelection.selectedChartId === chartId && !selectState) {
      synchronizedSelection.resetSelectionState?.();
    }
  }, [!!selectState]);
  const showVerticalCrosshair = hoveredChartId !== chartId && !synchronizedCrosshairDisabled;
  const showSynchronizedSelection = selectedChartId !== chartId && !synchronizedSelectionDisabled;
  const { currentScale: xScale } = (0, import_useXScale.useXScales)();
  const { datapointsMap, allDatapoints } = (0, import_useTimeseriesFilteredData.useTimeseriesFilteredData)();
  const yScale = (0, import_useYScales.useYScales)();
  const { position } = (0, import_selectors.useExploreState)();
  const zoomToSelection = (0, import_state_setters.useZoomToSelection)();
  const { closestPoint, filteredPoints, groupedPoints } = (0, import_get_points_parsed.getPointsParsed)(
    position?.relative,
    allDatapoints,
    xScale,
    yScale,
    datapointsMap,
    width,
    height
  );
  const handleModeChange = (event) => {
    const canPan = zoomState.zoomOutEnabled;
    const shouldSwitchToPanMode = event.mode === "pan" && pan?.enabled && canPan;
    const shouldSwitchToZoomMode = event.mode === "zoom" && (zoomX?.enabled || zoom?.enabled);
    if (shouldSwitchToPanMode || shouldSwitchToZoomMode) {
      setState((0, import_mode_reducers.reduceInteractionMode)(event.mode));
      setState(import_explore_reducers.reduceExploreReset);
      resetSynchronizedSelection?.();
    } else if (event.mode === "explore") {
      setState((0, import_mode_reducers.reduceInteractionMode)(import_types.INTERACTION_MODE.Explore));
    }
  };
  const handleHover = (payload) => {
    if (mode === "explore") {
      setState((0, import_explore_reducers.reduceExplorePosition)(payload.position));
    }
    const hoveredDate = xScale.invert(payload.position.relative.x);
    disableSynchronizedCrosshair?.(false);
    updateSynchronizedCrosshair?.(chartId, hoveredDate.valueOf());
  };
  const handleLeave = (_) => {
    if (mode === "explore") {
      setState(import_explore_reducers.reduceExploreEnd);
    }
    updateSynchronizedCrosshair?.(void 0);
  };
  const handlePanChange = (event) => {
    if (pan?.enabled) {
      applyPanDelta(event.delta);
    }
  };
  const handleTogglePin = (event) => {
    const highlightedSeriesId = closestPoint?.id && closestPoint.id !== import_constants.EMPTY_DATAPOINT.id ? closestPoint?.id : void 0;
    setState(
      event.pin ? (0, import_explore_reducers.reduceExplorePin)(event.position, highlightedSeriesId) : (0, import_explore_reducers.reduceExploreUnPin)(event.position)
    );
    updateSynchronizedCrosshair?.(void 0);
  };
  const handleReset = (event) => {
    disableSynchronizedCrosshair?.(false);
    resetSynchronizedSelection?.();
    resetInteractions(event.mousePosition);
  };
  const handleSelectStart = (event) => {
    if (mode === "explore") {
      updateSynchronizedSelection?.();
    } else if (!zoomX?.enabled) {
      handleTogglePin({
        triggeredBy: event.triggeredBy,
        position: event.position,
        pin: !exploreState.pinned
      });
    }
    disableSynchronizedCrosshair?.(true);
  };
  const handleSelectChange = (event) => {
    if (zoomX?.enabled || mode === "explore") {
      const variant = event.mode === "explore" ? "select" : "zoom";
      setState(
        (0, import_explore_reducers.reduceExploreSelect)({
          start: event.start,
          end: event.end,
          finished: false,
          accessible: false,
          variant,
          selectionPercentage: (0, import_get_percentage_from_pixel.getPercentageFromPixel)(
            [event.start, event.end],
            width
          )
        })
      );
      updateSynchronizedSelection?.({
        selection: [event.start, event.end],
        chartId,
        scale: xScale
      });
    }
  };
  const handleSelectEnd = (event) => {
    if ((0, import_types.isSelectCancelled)(event)) {
      setState(import_explore_reducers.reduceExploreCancelSelect);
      resetSynchronizedSelection?.();
    } else if (mode === "explore") {
      setState(
        (0, import_explore_reducers.reduceExploreSelect)({
          start: event.start,
          end: event.end,
          finished: true,
          accessible: false,
          selectionPercentage: (0, import_get_percentage_from_pixel.getPercentageFromPixel)(
            [event.start, event.end],
            width
          )
        })
      );
      selectionRef.current?.focus();
      updateSynchronizedSelection?.({
        selection: [event.start, event.end],
        chartId,
        scale: xScale
      });
    } else if (zoomX?.enabled) {
      zoomToSelection();
    }
  };
  const handleZoomChange = (event) => {
    setState(import_explore_reducers.reduceExploreCancelAll);
    applyZoomDelta(event.deltaStart, event.deltaEnd, event.behavior);
  };
  const handleToolbarMouseEnter = () => {
    setState(import_explore_reducers.reduceExploreCancelTooltip);
  };
  const handleFocus = (event) => {
    if (!event || event.cancelable) {
      return;
    }
    if ((0, import_is_own_event.isOwnEvent)(event)) {
      handlerRef.current.handleFocus();
    }
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_InteractionsOverlay.InteractionsOverlay,
    {
      "aria-label": import_constants2.TIMESERIES_INTERACTIONS_ARIA_LABEL,
      style: {
        zIndex: import_constants2.zIndicesTimeseriesAreas.overlays,
        userSelect: selectState?.finished === false ? "none" : "auto"
      },
      ref: overlayRef,
      mode,
      pinned: exploreState.pinned,
      selectionArea: selectState ? [selectState?.start, selectState?.end] : void 0,
      width,
      height,
      isZoomDisabled: !zoom?.enabled,
      isZoomXDisabled: !zoomX?.enabled,
      isPanDisabled: !(pan?.enabled && zoomState.zoomOutEnabled),
      onModeChange: handleModeChange,
      onFocus: handleFocus,
      onHover: handleHover,
      onLeave: handleLeave,
      onPanChange: handlePanChange,
      onTogglePin: handleTogglePin,
      onReset: handleReset,
      onSelectStart: handleSelectStart,
      onZoomChange: handleZoomChange,
      onSelectEnd: handleSelectEnd,
      onSelectChange: handleSelectChange,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_TimeseriesChartOverlayRefContext.TimeseriesChartOverlayRefContext.Provider, { value: overlayRef, children: [
        exploreState.visible ? /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_TimeseriesInspect.TimeseriesInspect,
            {
              filteredPoints,
              closestPoint,
              groupedPoints
            }
          ),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TimeseriesCrosshair.TimeseriesCrosshair, { closestPoint })
        ] }) : null,
        showVerticalCrosshair && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_VerticalSynchronizedCrosshair.VerticalSynchronizedCrosshair, { scale: xScale }),
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_TimeseriesChartSelectionAreaRef_context.TimeseriesChartSelectionAreaRef.Provider, { value: selectionRef, children: [
          toolbar?.enabled ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_TimeseriesChartToolbar.TimeseriesChartToolbar,
            {
              onMouseEnter: handleToolbarMouseEnter,
              handlerRef
            }
          ) : null,
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TimeseriesSelectionArea.TimeseriesSelectionArea, {})
        ] }),
        showSynchronizedSelection && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_SynchronizedSelectionArea.SynchronizedSelectionArea, { scale: xScale })
      ] })
    }
  );
};
TimeseriesChartOverlays["displayName"] = "TimeseriesChartOverlays";
