// @dynatrace/strato-components/charts/histogram/components/displays/BarChartDisplay.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var BarChartDisplay_exports = {};
__export(BarChartDisplay_exports, {
  BarChartDisplay: () => BarChartDisplay
});
module.exports = __toCommonJS(BarChartDisplay_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_accessors = require("./accessors.js");
var import_BarCanvas = require("../../../core/components/canvas/components/BarCanvas.js");
var import_useGridPlotSize = require("../../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_constants = require("../../../core/constants.js");
var import_useHistogramFilteredData = require("../../hooks/useHistogramFilteredData.js");
var import_useHistogramYScales = require("../../hooks/useHistogramYScales.js");
var import_HistogramDataProvider = require("../../providers/HistogramDataProvider.js");
function BarChartDisplay({
  xScale,
  series
}) {
  const yScales = (0, import_useHistogramYScales.useHistogramYScales)();
  const { width: widthPlot, height: heightPlot } = (0, import_useGridPlotSize.useGridPlotSize)();
  const { seriesMap, binsMap } = (0, import_useHistogramFilteredData.useHistogramFilteredData)();
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { "data-testid": "histogram-bar-chart-display", children: series.map((singleSeries) => {
    const seriesMetadata = (0, import_HistogramDataProvider.safeGet)(seriesMap, singleSeries);
    const yScale = yScales[seriesMetadata.scaleIndex];
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_BarCanvas.BarCanvas,
      {
        data: singleSeries.bins,
        canvasWidth: widthPlot,
        canvasHeight: heightPlot,
        x: (0, import_accessors.x)(xScale, binsMap),
        y: (0, import_accessors.y)(yScale, binsMap),
        width: (0, import_accessors.width)(xScale, binsMap),
        height: (0, import_accessors.height)(yScale, binsMap),
        strokeWidth: 1,
        color: () => seriesMetadata.color.Default,
        opacity: () => import_constants.DEFAULT_FILL_OPACITY
      },
      seriesMetadata.id
    );
  }) });
}
