// @dynatrace/strato-components/charts/timeseries/components/overlays/TimeseriesExplore/TimeseriesCrosshair.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TimeseriesCrosshair_exports = {};
__export(TimeseriesCrosshair_exports, {
  TimeseriesCrosshair: () => TimeseriesCrosshair
});
module.exports = __toCommonJS(TimeseriesCrosshair_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_useGridPlotSize = require("../../../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_HorizontalCrosshair = require("../../../../core/components/crosshair/HorizontalCrosshair.js");
var import_VerticalCrosshair = require("../../../../core/components/crosshair/VerticalCrosshair.js");
var import_constants = require("../../../../core/components/data-point/constants.js");
var import_reset_reducers = require("../../../../core/components/interactions-state-helpers/reducers/reset-reducers.js");
var import_constants2 = require("../../../../core/components/threshold-bar/types/constants.js");
var import_useApplyCallbackOnScroll = require("../../../../core/hooks/useApplyCallbackOnScroll.js");
var import_selectors = require("../../../hooks/store/selectors.js");
var import_useThresholdConfig = require("../../../hooks/useThresholdConfig.js");
var import_useTimeseriesFilteredData = require("../../../hooks/useTimeseriesFilteredData.js");
var import_useTimeseriesOverlayRef = require("../../../hooks/useTimeseriesOverlayRef.js");
var import_useTimeseriesXFormatter = require("../../../hooks/useTimeseriesXFormatter.js");
var import_useTimeseriesYFormatter = require("../../../hooks/useTimeseriesYFormatter.js");
var import_useXScale = require("../../../hooks/useXScale.js");
var import_useYScales = require("../../../hooks/useYScales.js");
var import_TimeseriesOverlays_css = require("../TimeseriesOverlays.sty.js");
var import_magnifiers_utils = require("../utils/magnifiers-utils.js");
function TimeseriesCrosshair({
  closestPoint
}) {
  const { units } = (0, import_useTimeseriesFilteredData.useTimeseriesFilteredData)();
  const yFormatters = (0, import_useTimeseriesYFormatter.useTimeseriesYFormatter)();
  const xFormatter = (0, import_useTimeseriesXFormatter.useTimeseriesXFormatter)();
  const yScale = (0, import_useYScales.useYScales)();
  const { currentScale: xScale } = (0, import_useXScale.useXScales)();
  const { width, height } = (0, import_useGridPlotSize.useGridPlotSize)();
  const overlaysRef = (0, import_useTimeseriesOverlayRef.useTimeseriesOverlayRef)();
  const { pinned, visible, position } = (0, import_selectors.useExploreState)();
  const { left: leftThreshold, right: rightThreshold } = (0, import_useThresholdConfig.useThresholdConfig)();
  const hasLeftThresholds = leftThreshold && leftThreshold.length > 0;
  const hasRightThresholds = rightThreshold && rightThreshold.length > 0;
  const relativeMousePosition = position?.relative;
  const chartBounds = overlaysRef?.current?.getBoundingClientRect();
  const isMouseHovering = relativeMousePosition && visible;
  const showVerticalCrosshair = isMouseHovering || pinned;
  const showHorizontalCrosshair = isMouseHovering && !pinned;
  const thresholdLeftSize = hasLeftThresholds ? import_constants2.THRESHOLD_BAR_THICKNESS : 0;
  const thresholdRightSize = hasRightThresholds ? import_constants2.THRESHOLD_BAR_THICKNESS : 0;
  const setState = (0, import_selectors.useSetState)();
  const isCallbackOnScrollDisabled = pinned || !isMouseHovering;
  const onScrollCallback = () => setState(import_reset_reducers.reduceResetExploreVisible);
  (0, import_useApplyCallbackOnScroll.useApplyCallbackOnScroll)(isCallbackOnScrollDisabled, onScrollCallback);
  const horizontalCrosshairWidth = width + thresholdLeftSize + thresholdRightSize;
  const verticalXCrosshair = pinned && closestPoint ? closestPoint.x - import_constants.OVERLAY_PADDING : relativeMousePosition?.x ?? 0;
  const yMagnifierLabels = relativeMousePosition ? (0, import_magnifiers_utils.generateYMagnifierLabels)(
    yScale,
    relativeMousePosition,
    units,
    yFormatters
  ) : void 0;
  const hoveredDate = xScale.invert(relativeMousePosition?.x ?? 0);
  const xMagnifierLabel = pinned ? "" : (0, import_magnifiers_utils.generateXMagnifierLabels)(xFormatter, hoveredDate);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("svg", { className: import_TimeseriesOverlays_css.crosshairLayout, children: [
    showHorizontalCrosshair && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_HorizontalCrosshair.HorizontalCrosshair,
      {
        leftLabel: yMagnifierLabels?.left,
        rightLabel: yMagnifierLabels?.right,
        y: relativeMousePosition?.y,
        width: horizontalCrosshairWidth,
        height,
        leftOffset: thresholdLeftSize,
        rightOffset: thresholdRightSize,
        chartBounds
      }
    ),
    showVerticalCrosshair && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_VerticalCrosshair.VerticalCrosshair,
      {
        x: verticalXCrosshair,
        bottomLabel: xMagnifierLabel,
        height,
        width: horizontalCrosshairWidth,
        dashed: !pinned,
        chartBounds
      }
    )
  ] });
}
