// @dynatrace/strato-components/charts/timeseries/components/overlays/TimeseriesExplore/TimeseriesChartTooltip.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TimeseriesChartTooltip_exports = {};
__export(TimeseriesChartTooltip_exports, {
  TimeseriesChartTooltip: () => TimeseriesChartTooltip
});
module.exports = __toCommonJS(TimeseriesChartTooltip_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_ChartActionsMenu = require("../../../../core/components/series-action/ChartActionsMenu.js");
var import_useChartActionsMenu = require("../../../../core/components/series-action/hooks/useChartActionsMenu.js");
var import_useDefaultAction = require("../../../../core/components/series-action/hooks/useDefaultAction.js");
var import_InteractiveChartTooltip = require("../../../../core/components/tooltip/InteractiveChartTooltip/components/InteractiveChartTooltip.js");
var import_SeriesBasedContent = require("../../../../core/components/tooltip/InteractiveChartTooltip/contents/SeriesBasedContent.js");
var import_TimebasedTooltipHeader = require("../../../../core/components/tooltip/InteractiveChartTooltip/headers/TimebasedTooltipHeader.js");
var import_useSeriesBasedItemEstimatedSize = require("../../../../core/components/tooltip/InteractiveChartTooltip/hooks/useSeriesBasedItemEstimatedSize.js");
var import_useTooltipSeriesActions = require("../../../../core/components/tooltip/shared/hooks/useTooltipSeriesActions.js");
var import_useScrollResponsiveChartBounds = require("../../../../core/hooks/useScrollResponsiveChartBounds.js");
var import_calculate_adjusted_tooltip_position = require("../../../../core/utils/calculate-adjusted-tooltip-position.js");
var import_safe_get = require("../../../../core/utils/safe-get.js");
var import_selectors = require("../../../hooks/store/selectors.js");
var import_useTimeseriesFilteredData = require("../../../hooks/useTimeseriesFilteredData.js");
var import_useTimeseriesOverlayRef = require("../../../hooks/useTimeseriesOverlayRef.js");
var import_useTimeseriesUnits = require("../../../hooks/useTimeseriesUnits.js");
var import_useTimeseriesYFormatter = require("../../../hooks/useTimeseriesYFormatter.js");
var import_useTooltipConfig = require("../../../hooks/useTooltipConfig.js");
var import_useTruncationMode = require("../../../hooks/useTruncationMode.js");
var import_constants = require("../constants.js");
var import_useSeriesActionClickHandler = require("../hooks/useSeriesActionClickHandler.js");
var import_sort_tooltip_items = require("../utils/tooltip/sort-tooltip-items.js");
var import_timeseries_tooltip = require("../utils/tooltip/timeseries-tooltip.js");
const TooltipPositionWrapper = ({
  tooltipItems,
  onMouseLeave,
  onScrollContent,
  children
}) => {
  const { pinned, visible, position } = (0, import_selectors.useExploreState)();
  const chartRef = (0, import_useTimeseriesOverlayRef.useTimeseriesOverlayRef)();
  const bounds = (0, import_useScrollResponsiveChartBounds.useScrollResponsiveChartBounds)(chartRef, pinned)();
  const { seriesDisplayMode } = (0, import_useTooltipConfig.useTooltipConfig)();
  const estimatedItemSize = (0, import_useSeriesBasedItemEstimatedSize.useSeriesBasedItemEstimatedSize)(
    tooltipItems,
    seriesDisplayMode
  );
  const adjustedTooltipPosition = (0, import_calculate_adjusted_tooltip_position.calculateAdjustedTooltipPosition)(
    bounds,
    position,
    { x: import_constants.ADJUSTED_X_AXIS, y: 0 }
  );
  const showTooltip = tooltipItems.length && visible && adjustedTooltipPosition;
  return showTooltip ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_InteractiveChartTooltip.InteractiveChartTooltip,
    {
      position: adjustedTooltipPosition,
      pinned,
      itemSize: estimatedItemSize,
      updatePositionWhenPinned: true,
      onMouseLeave,
      onScrollContent,
      children
    }
  ) : null;
};
const TimeseriesChartTooltip = ({
  series,
  closestPoint,
  xPoints,
  groupedPoints
}) => {
  const { seriesDisplayMode, variant } = (0, import_useTooltipConfig.useTooltipConfig)();
  const truncationMode = (0, import_useTruncationMode.useTruncationMode)();
  const { seriesMap, datapointsMap } = (0, import_useTimeseriesFilteredData.useTimeseriesFilteredData)();
  const {
    onLeave: handleMenuLeave,
    onScroll: handleMenuScroll,
    selectedItem,
    isScrolling,
    updateSelectedItem,
    actionsMenuRef
  } = (0, import_useChartActionsMenu.useChartActionsMenu)();
  const formatters = (0, import_useTimeseriesYFormatter.useTimeseriesYFormatter)();
  const units = (0, import_useTimeseriesUnits.useTimeseriesUnits)();
  const getTooltipTimeframe = () => [
    closestPoint.datapoint.start,
    closestPoint.datapoint.end
  ];
  const pointsToDisplay = variant === import_constants.SHARED_TOOLTIP_VARIANT ? (0, import_sort_tooltip_items.sortTooltipItems)(xPoints) : (0, import_sort_tooltip_items.sortTooltipItems)(
    (0, import_timeseries_tooltip.filterGroupedPoints)(closestPoint, xPoints, groupedPoints)
  );
  const tooltipItems = (0, import_timeseries_tooltip.toTooltipItems)(
    pointsToDisplay,
    formatters,
    datapointsMap,
    units
  );
  const closestItem = tooltipItems.find((item) => item.id === closestPoint.id);
  const actionClickHandler = (0, import_useSeriesActionClickHandler.useSeriesActionClickHandler)(
    tooltipItems.length > 1
  );
  const actions = (0, import_useTooltipSeriesActions.useTooltipSeriesActions)(
    series,
    (series2) => (0, import_safe_get.safeGet)(seriesMap, series2).id,
    tooltipItems.find((t) => t.id === selectedItem)?.datapoint
  );
  const isShared = variant === import_constants.SHARED_TOOLTIP_VARIANT;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    TooltipPositionWrapper,
    {
      tooltipItems,
      onMouseLeave: handleMenuLeave,
      onScrollContent: handleMenuScroll,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_InteractiveChartTooltip.InteractiveChartTooltip.Header, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_TimebasedTooltipHeader.TimebasedTooltipHeader,
          {
            timeframe: getTooltipTimeframe(),
            item: isShared && xPoints.length > 1 ? closestItem : void 0,
            seriesDisplayMode
          }
        ) }),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_InteractiveChartTooltip.InteractiveChartTooltip.Body, { children: tooltipItems.map((item, index) => () => {
          const isOpen = selectedItem === item.id;
          return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_ChartActionsMenu.ChartActionsMenu,
            {
              isOpen,
              actions: isOpen ? actions(item) : void 0,
              defaultActions: (0, import_useDefaultAction.getDefaultAction)(item),
              ref: actionsMenuRef,
              onActionClick: actionClickHandler,
              children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_SeriesBasedContent.SeriesBasedContent,
                {
                  item: { ...item, values: item.asyncValues() },
                  onMouseHover: () => updateSelectedItem(item.id),
                  seriesDisplayMode,
                  showDivider: index !== tooltipItems.length - 1,
                  truncationMode,
                  isScrolling
                }
              )
            },
            `${item.id}`
          );
        }) })
      ]
    }
  );
};
