// @dynatrace/strato-components/tables/DataTable/components/TableActions/DataTableSelectedRowsActions.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DataTableSelectedRowsActions_exports = {};
__export(DataTableSelectedRowsActions_exports, {
  DataTableSelectedRowsActions: () => DataTableSelectedRowsActions,
  _isDataTableSelectedRowsActions: () => _isDataTableSelectedRowsActions
});
module.exports = __toCommonJS(DataTableSelectedRowsActions_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_useBehavioralTrackingProps = require("../../../../core/hooks/useBehavioralTrackingProps.js");
var import_logging = require("../../../../core/utils/logging.js");
var import_Flex = require("../../../../layouts/flex/Flex.js");
var import_DataTableContext = require("../../contexts/DataTableContext.js");
function _isDataTableSelectedRowsActions(child) {
  return child?.type === DataTableSelectedRowsActions;
}
const DataTableSelectedRowsActions = (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const table = (0, import_DataTableContext.useDataTable)();
    const {
      children,
      "data-testid": dataTestId,
      className: consumerClassName,
      style: consumerStyle,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
    if (typeof children !== "function") {
      (0, import_logging.devLog)("The SelectedRowsActions sub-component child must be a function!");
      return null;
    }
    const currentSelectableRows = table.getSelectedRowModel().flatRows.map((row) => ({
      rowId: row.id,
      originalRow: row.original
    }));
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Flex.Flex,
      {
        ref: forwardedRef,
        justifyContent: "flex-start",
        alignItems: "flex-end",
        flexWrap: "wrap",
        "data-testid": dataTestId,
        className: consumerClassName,
        style: consumerStyle,
        ...behavioralTrackingProps,
        ...restProps,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Flex.Flex, { alignItems: "center", flexWrap: "wrap", gap: 4, children: children(currentSelectableRows) })
      }
    );
  }
);
DataTableSelectedRowsActions.displayName = "DataTable.SelectedRowsActions";
