// @dynatrace/strato-components/tables/DataTable/components/TableActions/DataTableTableActions.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DataTableTableActions_exports = {};
__export(DataTableTableActions_exports, {
  DataTableTableActions: () => DataTableTableActions,
  _isTableActions: () => _isTableActions
});
module.exports = __toCommonJS(DataTableTableActions_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_useBehavioralTrackingProps = require("../../../../core/hooks/useBehavioralTrackingProps.js");
var import_Flex = require("../../../../layouts/flex/Flex.js");
function _isTableActions(child) {
  return child?.type === DataTableTableActions;
}
const DataTableTableActions = (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const {
      children,
      "data-testid": dataTestId,
      className: consumerClassName,
      style: consumerStyle,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Flex.Flex,
      {
        ref: forwardedRef,
        justifyContent: "flex-start",
        alignItems: "flex-end",
        flexWrap: "wrap",
        gap: 4,
        "data-testid": dataTestId,
        className: consumerClassName,
        style: consumerStyle,
        ...behavioralTrackingProps,
        ...restProps,
        children
      }
    );
  }
);
DataTableTableActions.displayName = "DataTable.TableActions";
