// @dynatrace/strato-components/esm/charts/timeseries/hooks/store/selectors.js
import { isEqual } from "lodash-es";
import { useContext } from "react";
import { useStoreSelector } from "../../../../core/store/selector.js";
import { TimeseriesChartStateContext } from "../../context/TimeseriesChartState.context.js";
const useSelector = (selector, compareFn) => useStoreSelector(TimeseriesChartStateContext, selector, compareFn);
const useZoomState = () => useSelector((state) => state.zoom);
const useAnnotationsState = () => useSelector((state) => state.annotations);
const useThresholdsState = () => useSelector((state) => state.thresholds);
const useFilteredSeriesIdsState = () => useSelector((state) => state.filteredSeriesIds);
const useHighlightedSeriesIdState = () => useSelector((state) => state.highlightedSeriesId);
const useCurrentDomainState = () => useSelector((state) => state.currentDomain);
const useInteractionModeState = () => useSelector((state) => state.interactionMode);
const useExploreState = () => useSelector((state) => state.explore, isEqual);
const useSelectState = () => useSelector((state) => state.select);
const useSetState = () => {
  const { setState } = useContext(TimeseriesChartStateContext);
  return setState;
};
export {
  useAnnotationsState,
  useCurrentDomainState,
  useExploreState,
  useFilteredSeriesIdsState,
  useHighlightedSeriesIdState,
  useInteractionModeState,
  useSelectState,
  useSelector,
  useSetState,
  useThresholdsState,
  useZoomState
};
