// @dynatrace/strato-components/charts/core/components/legend/categorical/components/ChartLegendGrid.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ChartLegendGrid_exports = {};
__export(ChartLegendGrid_exports, {
  ChartLegendGrid: () => ChartLegendGrid
});
module.exports = __toCommonJS(ChartLegendGrid_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_ChartLegendItem = require("./ChartLegendItem/ChartLegendItem.js");
var import_ChartVirtualizedLegendGrid_css = require("./ChartVirtualizedLegendGrid.sty.js");
var import_constants = require("../../../../constants.js");
var import_useChartLayoutRatio = require("../../../chart-layout/hooks/useChartLayoutRatio.js");
var import_constants2 = require("../constants.js");
var import_LegendGridSettingsContext = require("../context/LegendGridSettingsContext.js");
const ChartLegendGrid = (props) => {
  const { items, orientation } = props;
  const { legendRef } = (0, import_useChartLayoutRatio.useChartLayoutRatio)();
  const {
    maxRows,
    minItemWidth,
    canFitAllItemsWithoutScrollbar,
    maxItemWidth
  } = (0, import_LegendGridSettingsContext.useLegendGridSettings)();
  const nrOfColumns = Math.ceil(items.length / maxRows);
  const gridTemplateColumns = !canFitAllItemsWithoutScrollbar || orientation === import_constants.ORIENTATION.Horizontal ? `repeat(auto-fit, minmax(${minItemWidth - import_constants2.LEGEND_ACTION_BUTTON_WIDTH}px, 1fr))` : `repeat(${nrOfColumns}, minmax(${minItemWidth}px, ${maxItemWidth}px))`;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: (0, import_ChartVirtualizedLegendGrid_css.legendContent)({ layout: orientation }), children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      ref: legendRef,
      "data-testid": "chart-legend-grid-static",
      className: (0, import_clsx.default)(
        (0, import_ChartVirtualizedLegendGrid_css.legendGrid)({ layout: orientation }),
        import_constants2.CATEGORICAL_LEGEND_CLASS
      ),
      style: { gridTemplateColumns },
      children: items.map((item) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_ChartLegendItem.ChartLegendItem,
        {
          item,
          numberOfItems: items.length
        },
        item.id
      ))
    }
  ) });
};
ChartLegendGrid["displayName"] = "ChartLegendGrid";
