// @dynatrace/strato-components/charts/core/components/legend/categorical/components/ChartVirtualizedLegendGrid.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ChartVirtualizedLegendGrid_exports = {};
__export(ChartVirtualizedLegendGrid_exports, {
  ChartVirtualizedLegendGrid: () => ChartVirtualizedLegendGrid
});
module.exports = __toCommonJS(ChartVirtualizedLegendGrid_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_virtual = require("@tanstack/react-virtual");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_ChartLegendItem = require("./ChartLegendItem/ChartLegendItem.js");
var import_ChartVirtualizedLegendGrid_css = require("./ChartVirtualizedLegendGrid.sty.js");
var import_useChartLayoutRatio = require("../../../chart-layout/hooks/useChartLayoutRatio.js");
var import_LegendSizeContext = require("../../shared/context/LegendSizeContext.js");
var import_constants = require("../constants.js");
var import_LegendGridSettingsContext = require("../context/LegendGridSettingsContext.js");
const ChartVirtualizedLegendGrid = (props) => {
  const { items, orientation } = props;
  const { legendRef } = (0, import_useChartLayoutRatio.useChartLayoutRatio)();
  const { minItemWidth } = (0, import_LegendGridSettingsContext.useLegendGridSettings)();
  const { width: legendWidth } = (0, import_LegendSizeContext.useLegendSize)();
  const seriesCount = items.length;
  const nrOfColumns = Math.max(Math.floor(legendWidth / minItemWidth), 1);
  const nrOfRows = Math.ceil(seriesCount / nrOfColumns);
  const parentRef = (0, import_react.useRef)(null);
  const virtualizer = (0, import_react_virtual.useVirtualizer)({
    count: nrOfRows,
    getScrollElement: () => parentRef.current,
    estimateSize: (0, import_react.useCallback)(() => import_constants.ITEM_HEIGHT.value, []),
    overscan: 0
  });
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      ref: parentRef,
      className: `${(0, import_ChartVirtualizedLegendGrid_css.legendContent)({
        layout: orientation
      })} chart-legend-grid-content`,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        "div",
        {
          ref: legendRef,
          "data-testid": "chart-legend-grid-virtual",
          className: (0, import_clsx.default)(import_ChartVirtualizedLegendGrid_css.virtualizedGrid, import_constants.CATEGORICAL_LEGEND_CLASS),
          style: {
            height: virtualizer.getTotalSize()
          },
          children: virtualizer.getVirtualItems().map((virtualRow) => {
            const rowStartIndex = virtualRow.index * nrOfColumns;
            const rowEndIndex = rowStartIndex + nrOfColumns;
            const gridTemplateColumns = `repeat(${nrOfColumns}, 1fr)`;
            return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              "div",
              {
                className: import_ChartVirtualizedLegendGrid_css.virtualizedRow,
                style: {
                  height: `${virtualRow.size}px`,
                  transform: `translateY(${virtualRow.start}px)`
                },
                children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  "div",
                  {
                    className: (0, import_ChartVirtualizedLegendGrid_css.legendGrid)({ layout: orientation }),
                    style: { gridTemplateColumns },
                    children: items.slice(rowStartIndex, rowEndIndex).map((item) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                      import_ChartLegendItem.ChartLegendItem,
                      {
                        item,
                        numberOfItems: items.length
                      },
                      item.id
                    ))
                  }
                )
              },
              virtualRow.index
            );
          })
        }
      )
    }
  );
};
