// @dynatrace/strato-components/layouts/container/Container.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Container_exports = {};
__export(Container_exports, {
  Container: () => Container,
  useContainerColoring: () => useContainerColoring
});
module.exports = __toCommonJS(Container_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_Container_css = require("./Container.sty.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_container_css = require("../../styles/container.sty.js");
var import_getLayoutSizeStyles = require("../../styles/getLayoutSizeStyles.js");
var import_getSpacingSprinkles = require("../../styles/getSpacingSprinkles.js");
const COMPONENT_NAME = "Container";
const ContainerColoringContext = (0, import_react.createContext)({
  variant: "default",
  color: "neutral",
  default: true
});
const useContainerColoring = () => {
  return (0, import_react.useContext)(ContainerColoringContext);
};
const Container = /* @__PURE__ */ (0, import_react.forwardRef)(
  ({
    as,
    children,
    variant = "default",
    color = "neutral",
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "data-dt-component": dataDtComponent = COMPONENT_NAME,
    ...remainingProps
  }, ref) => {
    const [spacingClasses, sprinkleRestProps] = (0, import_getSpacingSprinkles.getSpacingSprinkles)(
      remainingProps,
      { padding: 16 }
    );
    const [sizeStyles, sizeRestProps] = (0, import_getLayoutSizeStyles.getLayoutSizeStyles)(sprinkleRestProps);
    const [behavioralTrackingProps, finalRestProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(sizeRestProps);
    const Component = as || "div";
    const context = (0, import_react.useMemo)(
      () => ({ color, variant, default: false }),
      [color, variant]
    );
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(ContainerColoringContext.Provider, { value: context, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      Component,
      {
        ref,
        className: (0, import_clsx.default)(
          (0, import_container_css.containerColorsCSS)({ color, variant }),
          import_Container_css.containerCSS,
          spacingClasses,
          consumerClassName
        ),
        style: {
          ...sizeStyles,
          ...consumerStyle
        },
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        ...finalRestProps,
        ...behavioralTrackingProps,
        "data-dt-component": dataDtComponent,
        children
      }
    ) });
  }
);
Container.displayName = "Container";
