// @dynatrace/strato-components/esm/layouts/app-header/action-group/ActionButton.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import {
  forwardRef
} from "react";
import * as styles from "./ActionButton.sty.js";
import { useActionGroupContext } from "./hooks/useActionGroupContext.js";
import { Button } from "../../../buttons/button/Button.js";
import { focusRingStatic } from "../../../core/styles/focusRingStatic.sty.js";
import { Menu } from "../../../navigation/menu/Menu.js";
import { textStyle as textStyleCSS } from "../../../styles/textStyle.sty.js";
const ActionButton = forwardRef(
  (props, forwardedRef) => {
    const {
      showLabel = true,
      isSelected,
      children,
      prefixIcon,
      disabled,
      as,
      pin,
      onClick,
      density = "default",
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      ...remainingProps
    } = props;
    const { isCollapsed } = useActionGroupContext();
    const showChildren = !prefixIcon || showLabel || isCollapsed;
    const sharedProps = {
      disabled,
      className: clsx(
        consumerClassName,
        styles.navItem({
          density
        }),
        styles.navItemSelectionStyle,
        {
          [styles.disabled]: disabled,
          active: isSelected
        }
      ),
      "data-testid": dataTestId
    };
    if (isCollapsed) {
      return /* @__PURE__ */ jsxs(
        Menu.Item,
        {
          ref: forwardedRef,
          as,
          onSelect: onClick ?? (() => null),
          ...sharedProps,
          style: consumerStyle,
          ...remainingProps,
          children: [
            prefixIcon && /* @__PURE__ */ jsx(Menu.Prefix, { children: prefixIcon }),
            showChildren && children
          ]
        }
      );
    }
    const sharedActionButtonProps = {
      ...sharedProps,
      onClick,
      className: clsx(
        focusRingStatic({ disabled }),
        textStyleCSS({ textStyle: "base-emphasized" }),
        sharedProps.className
      )
    };
    if (as) {
      const Component = as;
      return /* @__PURE__ */ jsxs(
        Component,
        {
          ref: forwardedRef,
          ...sharedActionButtonProps,
          style: consumerStyle,
          ...remainingProps,
          children: [
            prefixIcon,
            showChildren && children
          ]
        }
      );
    }
    return /* @__PURE__ */ jsxs(
      Button,
      {
        ref: forwardedRef,
        ...sharedActionButtonProps,
        style: {
          display: "flex",
          minWidth: 0,
          flexShrink: 0,
          ...consumerStyle
        },
        ...remainingProps,
        children: [
          /* @__PURE__ */ jsx(Button.Prefix, { children: prefixIcon }),
          showChildren && children
        ]
      }
    );
  }
);
ActionButton.displayName = "ActionGroup.Button";
export {
  ActionButton
};
