// @dynatrace/strato-components/esm/layouts/app-header/action-group/ActionGroup.js
import { jsx, jsxs } from "react/jsx-runtime";
import { createElement } from "react";
import clsx from "clsx";
import {
  Children,
  cloneElement,
  forwardRef,
  Fragment,
  useEffect,
  useMemo,
  useRef,
  useState
} from "react";
import { isElement } from "react-is";
import { ActionButton } from "./ActionButton.js";
import { actionGroupHiddenItemCSS } from "./ActionGroup.sty.js";
import { CollapsedActionGroup } from "./CollapsedActionGroup.js";
import { ACTION_GROUP_GAP } from "./constants.js";
import { ActionGroupContext } from "./contexts/ActionGroupContext.js";
import { Group } from "./Group.js";
import { NavItem } from "./NavItem.js";
import { useContainerSize } from "../../../core/hooks/useContainerSize.js";
import { useMeasureChildren } from "../../../core/hooks/useMeasureChildren.js";
import { useMergeRefs } from "../../../core/hooks/useMergeRefs.js";
import { useRovingFocus } from "../../../core/hooks/useRovingFocus.js";
import { mergeProps as _mergeProps } from "../../../core/utils/merge-props.js";
import { Divider } from "../../divider/Divider.js";
import { Flex } from "../../flex/Flex.js";
const COMPONENT_NAME = "ActionGroup";
function itemIsPinned(child) {
  return isElement(child) && typeof child.props === "object" && child.props !== null && "pin" in child.props && child.props.pin === true;
}
const ActionGroupRoot = forwardRef(
  (props, forwardedRef) => {
    const {
      collapsible = true,
      collapseFrom = "start",
      children,
      rovingFocus = true,
      triggerIcon,
      density = "default",
      onCollapsedChange,
      onOverflowChange,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      ...restProps
    } = props;
    const _containerRef = useRef(null);
    const { containerRef, measurements, childProps } = useMeasureChildren();
    const measuredChildren = useRef([]);
    const { width } = useContainerSize(_containerRef);
    const containerWidth = Math.round(width);
    const menuItems = useRef(/* @__PURE__ */ new Map());
    const pinnedItemsCount = useRef(0);
    const [isCollapsed, setIsCollapsed] = useState(false);
    const [isOverflowing, setIsOverflowing] = useState(false);
    const totalChildrenWidth = useMemo(() => {
      let childWidths = Object.values(measurements).reduce(
        (acc, value) => acc += value?.width || 0,
        0
      );
      childWidths += (childProps.length - 1) * ACTION_GROUP_GAP;
      return childWidths;
    }, [childProps, measurements]);
    const slots = useMemo(() => {
      const slots2 = { items: [], menuItems: [] };
      const childrenArray = Children.toArray(children);
      if (collapsible && containerWidth > 0 && childProps.length > 0 && Object.keys(measurements).length === childProps.length) {
        const childKeys = childProps.map((props2) => props2["data-id"]);
        let childIndex = collapseFrom === "start" ? 0 : childProps.length - 1;
        menuItems.current.clear();
        let childWidths = totalChildrenWidth;
        while (childWidths > containerWidth && (collapseFrom === "end" && childIndex >= 0 || collapseFrom === "start" && childIndex < childProps.length)) {
          const childNode = measuredChildren.current[
            // The childProps array also includes props for the menu trigger, which is not included in the
            // measured children array. If collapsing from the end, the index needs to be shifted back by 1.
            childIndex + (isOverflowing && collapseFrom === "end" ? -1 : 0)
          ];
          if (isElement(childNode) && itemIsPinned(childNode)) {
          } else {
            let childKeyIndex = childIndex;
            if (isOverflowing) {
              const collapseFromOffset = collapseFrom === "start" ? 1 : -1;
              childKeyIndex = childKeyIndex + collapseFromOffset;
            }
            const key = childKeys[childKeyIndex];
            if (key && measurements[key] !== void 0) {
              menuItems.current.set(key, true);
              childWidths -= (measurements[key]?.width || 0) + ACTION_GROUP_GAP;
            }
          }
          childIndex += collapseFrom === "start" ? 1 : -1;
        }
      }
      let groupCount = 0;
      let pinnedItems = 0;
      measuredChildren.current = [];
      for (const [index, child] of childrenArray.entries()) {
        if (!isElement(child)) {
          continue;
        }
        const childKey = index + groupCount;
        const childIndex = childKey + (isOverflowing && collapseFrom === "start" ? 1 : 0);
        const isItemOverflowing = menuItems.current.get(
          childProps[childIndex]?.["data-id"] || ""
        );
        const clonedChild = cloneElement(
          child,
          _mergeProps(child.props ?? {}, childProps[childIndex], {
            key: childKey,
            className: clsx(props.className, {
              [actionGroupHiddenItemCSS]: isItemOverflowing
            })
          })
        );
        if (child.type === Group) {
          const isNextItemOverflowing = menuItems.current.get(
            childProps[childIndex + 2]?.["data-id"] || ""
          );
          measuredChildren.current.push(clonedChild);
          slots2.items.push(
            /* @__PURE__ */ jsxs(Fragment, { children: [
              clonedChild,
              index !== childrenArray.length - 1 && /* @__PURE__ */ createElement(
                Divider,
                {
                  orientation: "vertical",
                  flexItem: false,
                  style: {
                    height: 20
                  },
                  className: clsx({
                    [actionGroupHiddenItemCSS]: !!isItemOverflowing || !!isNextItemOverflowing
                  }),
                  ...childProps[childIndex + 1],
                  key: childKey + 1
                }
              )
            ] }, `fragment-${childKey}`)
          );
          if (itemIsPinned(child)) {
            measuredChildren.current.push({ props: { pin: true } });
            pinnedItems += 1;
          } else {
            measuredChildren.current.push({});
          }
          groupCount += 1;
        } else {
          measuredChildren.current.push(clonedChild);
          slots2.items.push(clonedChild);
          if (itemIsPinned(child)) {
            pinnedItems += 1;
          }
        }
        if (isItemOverflowing) {
          slots2.menuItems.push(child);
        }
      }
      pinnedItemsCount.current = pinnedItems;
      setIsOverflowing(slots2.menuItems.length > 0);
      setIsCollapsed(
        slots2.menuItems.length + pinnedItemsCount.current === slots2.items.length
      );
      return slots2;
    }, [
      measurements,
      childProps,
      children,
      collapseFrom,
      collapsible,
      totalChildrenWidth,
      containerWidth
    ]);
    const { containerRef: rovingContainerRef } = useRovingFocus(
      {
        // FIXME: Follow-up for detecting the initial index already, when it is initially undefined based on
        // either tab-index attributes or the first focusable element. This will solve the problem for
        // everyone and will not overcomplicate the logic. ADES-2260
        initialIndex: void 0,
        direction: "horizontal",
        focusableChildrenSelector: ':scope a, :scope button, :scope [type="button"]'
      }
    );
    const rootRef = useMergeRefs([
      forwardedRef,
      ...rovingFocus ? [rovingContainerRef] : []
    ]);
    const collapsibleRootRef = useMergeRefs([
      rootRef,
      containerRef,
      _containerRef
    ]);
    useEffect(() => {
      onOverflowChange?.(isOverflowing);
    }, [isOverflowing, onOverflowChange]);
    useEffect(() => {
      onCollapsedChange?.(isCollapsed);
    }, [isCollapsed, onCollapsedChange]);
    return /* @__PURE__ */ jsx(
      Flex,
      {
        "data-actiongroup": true,
        alignItems: "center",
        "data-testid": dataTestId,
        ref: collapsible ? collapsibleRootRef : rootRef,
        gap: ACTION_GROUP_GAP,
        className: consumerClassName,
        style: { minWidth: !collapsible ? 0 : void 0, ...consumerStyle },
        ...restProps,
        "data-dt-component": COMPONENT_NAME,
        children: /* @__PURE__ */ jsxs(
          ActionGroupContext.Provider,
          {
            value: {
              isCollapsed: false
            },
            children: [
              collapseFrom === "start" && slots?.menuItems.length > 0 && /* @__PURE__ */ jsx(
                CollapsedActionGroup,
                {
                  triggerIcon,
                  childProps: childProps[0],
                  density,
                  children: slots?.menuItems
                }
              ),
              slots?.items,
              collapseFrom === "end" && slots?.menuItems.length > 0 && /* @__PURE__ */ jsx(
                CollapsedActionGroup,
                {
                  triggerIcon,
                  childProps: childProps[childProps.length - 1],
                  density,
                  children: slots?.menuItems
                }
              )
            ]
          }
        )
      }
    );
  }
);
ActionGroupRoot.displayName = "ActionGroup";
const ActionGroup = Object.assign(ActionGroupRoot, {
  ActionButton,
  Group,
  NavItem
});
export {
  ActionGroup
};
