// @dynatrace/strato-components/esm/layouts/app-header/action-group/CollapsedActionGroup.js
import { jsx, jsxs } from "react/jsx-runtime";
import { forwardRef, useState } from "react";
import { useIntl } from "react-intl";
import { DotMenuIcon } from "@dynatrace/strato-icons";
import { ActionGroupContext } from "./contexts/ActionGroupContext.js";
import { Button } from "../../../buttons/button/Button.js";
import { Menu } from "../../../navigation/menu/Menu.js";
const CollapsedActionGroup = forwardRef((props, forwardedRef) => {
  const {
    children,
    density = "default",
    triggerIcon = /* @__PURE__ */ jsx(DotMenuIcon, {}),
    childProps
  } = props;
  const intl = useIntl();
  const [open, setOpen] = useState(false);
  return /* @__PURE__ */ jsx(ActionGroupContext.Provider, { value: { isCollapsed: true }, children: /* @__PURE__ */ jsxs(Menu, { open, onOpenChange: setOpen, children: [
    /* @__PURE__ */ jsx(Menu.Trigger, { children: /* @__PURE__ */ jsx(
      Button,
      {
        variant: "default",
        size: density,
        ...childProps,
        "aria-label": intl.formatMessage({
          id: "L/D6X7Ee5PYWzl3i",
          defaultMessage: "Show more actions"
        }),
        children: /* @__PURE__ */ jsx(Button.Prefix, { children: triggerIcon })
      }
    ) }),
    open && /* @__PURE__ */ jsx(Menu.Content, { ref: forwardedRef, children })
  ] }) });
});
export {
  CollapsedActionGroup
};
