// @dynatrace/strato-components/esm/layouts/app-header/action-group/Group.js
import { jsx } from "react/jsx-runtime";
import { forwardRef, Children } from "react";
import { isElement } from "react-is";
import { ACTION_GROUP_GAP } from "./constants.js";
import { useActionGroupContext } from "./hooks/useActionGroupContext.js";
import { Menu } from "../../../navigation/menu/Menu.js";
import { Flex } from "../../flex/Flex.js";
const Group = forwardRef(
  (props, forwardedRef) => {
    const {
      children,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      ...remainingProps
    } = props;
    const slots = {
      items: []
    };
    for (const child of Children.toArray(children)) {
      if (isElement(child)) {
        slots.items?.push(child);
      }
    }
    const { isCollapsed } = useActionGroupContext();
    return isCollapsed ? /* @__PURE__ */ jsx(
      Menu.Group,
      {
        ref: forwardedRef,
        className: consumerClassName,
        style: consumerStyle,
        "data-testid": dataTestId,
        children: slots.items
      }
    ) : /* @__PURE__ */ jsx(
      Flex,
      {
        ref: forwardedRef,
        gap: ACTION_GROUP_GAP,
        alignItems: "center",
        role: "group",
        className: consumerClassName,
        style: { ...consumerStyle, flexShrink: 0 },
        "data-testid": dataTestId,
        ...remainingProps,
        children: slots.items
      }
    );
  }
);
Group.displayName = "ActionGroup.Group";
export {
  Group
};
