// @dynatrace/strato-components/esm/layouts/app-header/action-group/NavItem.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import { forwardRef } from "react";
import * as styles from "./NavItem.sty.js";
import { Button } from "../../../buttons/button/Button.js";
const NavItem = /* @__PURE__ */ forwardRef(
  (props, forwardedRef) => {
    const {
      disabled = false,
      isSelected = false,
      as = "a",
      className: consumerClassName,
      style: consumerStyle = {},
      "data-testid": dataTestId,
      ...remainingProps
    } = props;
    return /* @__PURE__ */ jsx(
      Button,
      {
        ...remainingProps,
        ref: forwardedRef,
        disabled,
        "data-testid": dataTestId,
        as,
        className: clsx(
          styles.button,
          {
            [styles.isSelected]: isSelected,
            [styles.isDisabled]: disabled
          },
          consumerClassName
        ),
        style: {
          minWidth: 0,
          flexShrink: 0,
          display: "flex",
          ...consumerStyle
        }
      }
    );
  }
);
NavItem.displayName = "AppHeader.NavItem";
export {
  NavItem
};
