// @dynatrace/strato-components/charts/core/components/single-value-renderer/providers/font-size.provider.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var font_size_provider_exports = {};
__export(font_size_provider_exports, {
  FontSizeProvider: () => FontSizeProvider
});
module.exports = __toCommonJS(font_size_provider_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_useId = require("../../../../../core/hooks/useId.js");
var import_useMergeRefs = require("../../../../../core/hooks/useMergeRefs.js");
var import_useFontSizeForContainer = require("../../../hooks/useFontSizeForContainer.js");
var import_useThrottledResizeObserver = require("../../../hooks/useThrottledResizeObserver.js");
var import_constants = require("../constants.js");
var import_font_size_context = require("../contexts/font-size.context.js");
var import_use_global_font_size = require("../hooks/use-global-font-size.js");
const FontSizeProvider = ({
  text,
  suffix,
  children
}) => {
  const ref = (0, import_react.useRef)(null);
  const { refCallback, width = 0, height = 0 } = (0, import_useThrottledResizeObserver.useThrottledResizeObserver)();
  const mergedRef = (0, import_useMergeRefs.useMergeRefs)([refCallback, ref]);
  let preferredFontSize = (0, import_useFontSizeForContainer.useFontSizeForContainer)({
    container: {
      width: width * import_constants.FONT_WIDTH_RATIO,
      height: height * import_constants.FONT_HEIGHT_RATIO
    },
    text,
    element: ref.current?.querySelector("span"),
    // element containing text font style
    suffix,
    suffixElement: ref.current?.querySelector("small")
    // element containing suffix font style
  });
  if (text.length === 0 || text === "00") {
    preferredFontSize = import_constants.DEFAULT_MAX_VALUE_FONT_SIZE;
  }
  const id = (0, import_useId.useId)();
  const { updateValueFontSize } = (0, import_use_global_font_size.useGlobalFontSize)();
  (0, import_react.useLayoutEffect)(() => {
    if (height && width) {
      updateValueFontSize?.(id, preferredFontSize);
    }
  }, [height, width, preferredFontSize]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_font_size_context.FontSizeContext.Provider,
    {
      value: {
        fontSize: preferredFontSize,
        refCallback: mergedRef
      },
      children
    }
  );
};
