// @dynatrace/strato-geo/esm/map/utils/iterate-config-slots.js
import { isEmpty, isUndefined } from "lodash-es";
import { Children } from "react";
import { ChartToolbar } from "@dynatrace/strato-components/charts";
import {
  DEFAULT_BASE_LAYER_RULES,
  DEFAULT_MAP_LEGEND_RATIO,
  DEFAULT_MAP_POSITION,
  DEFAULT_SEQUENTIAL_LEGEND_COLOR_PALETTE
} from "../constants.js";
import { isSlot } from "./is-slot.js";
import { replaceInfiniteValuesInRanges } from "./replace-infinite-values-in-ranges.js";
import { BaseLayer } from "../slots/BaseLayer.js";
import { CategoricalLegend } from "../slots/CategoricalLegend.js";
import { ChartInteractions } from "../slots/ChartInteractions.js";
import { DownloadCSV } from "../slots/DownloadCSV.js";
import { SequentialLegend } from "../slots/SequentialLegend.js";
import { ThresholdLegend } from "../slots/ThresholdLegend.js";
import { sanitizeProps, Toolbar } from "../slots/Toolbar.js";
import { Zoom } from "../slots/Zoom.js";
const iterateConfigSlots = (children, legendDomain) => {
  let slots = {
    toolbar: void 0,
    interactions: void 0,
    legend: void 0,
    baseLayer: DEFAULT_BASE_LAYER_RULES
  };
  let slotMapToolbarConfig = {};
  Children.forEach(children, (child) => {
    if (isSlot(child, ChartInteractions)) {
      const { children: children2, ...previousInteractions } = child.props;
      const interactions = iterateChartInteractionsSlots(children2);
      slots = {
        ...slots,
        interactions: {
          ...previousInteractions,
          ...interactions
        }
      };
    } else if (isSlot(child, ChartToolbar)) {
      const { children: children2, ...toolbar } = child.props;
      const childrenToolbar = iterateChartToolbarSlots(children2);
      slots = {
        ...slots,
        toolbar: {
          ...toolbar,
          ...childrenToolbar
        }
      };
    } else if (isSlot(child, SequentialLegend)) {
      slots = {
        ...slots,
        legend: {
          type: "sequential",
          colorPalette: DEFAULT_SEQUENTIAL_LEGEND_COLOR_PALETTE,
          position: DEFAULT_MAP_POSITION,
          ratio: DEFAULT_MAP_LEGEND_RATIO,
          ...child.props,
          min: child.props.min ?? legendDomain[0],
          max: child.props.max ?? legendDomain[1]
        }
      };
    } else if (isSlot(child, CategoricalLegend)) {
      slots = {
        ...slots,
        legend: {
          type: "categorical",
          position: DEFAULT_MAP_POSITION,
          ...child.props
        }
      };
    } else if (isSlot(child, ThresholdLegend)) {
      const getDomainValue = (index, legendDomainValue) => Number.isFinite(legendDomainValue) ? legendDomainValue : child.props.ranges?.at(index)?.to ?? legendDomainValue;
      const ranges = replaceInfiniteValuesInRanges(child.props.ranges, [
        getDomainValue(0, legendDomain[0]),
        getDomainValue(-1, legendDomain[1])
      ]);
      slots = {
        ...slots,
        legend: {
          type: "threshold",
          position: DEFAULT_MAP_POSITION,
          ratio: DEFAULT_MAP_LEGEND_RATIO,
          ...child.props,
          ranges
        }
      };
    } else if (isSlot(child, Toolbar)) {
      slotMapToolbarConfig = {
        ...slotMapToolbarConfig,
        ...sanitizeProps(child.props)
      };
    } else if (isSlot(child, DownloadCSV)) {
      slotMapToolbarConfig = {
        ...slotMapToolbarConfig,
        downloadCSV: { enabled: true }
      };
    } else if (isSlot(child, Zoom)) {
      slotMapToolbarConfig = {
        ...slotMapToolbarConfig,
        zoom: { enabled: child.props.disabled !== true }
      };
    } else if (isSlot(child, BaseLayer)) {
      slots = {
        ...slots,
        baseLayer: {
          ...slots.baseLayer,
          ...child.props
        }
      };
    }
  });
  const shouldEnableDefaultZoom = isUndefined(slotMapToolbarConfig.zoom) && !isEmpty(slotMapToolbarConfig) && slots.toolbar?.hidden !== true;
  if (shouldEnableDefaultZoom) {
    slotMapToolbarConfig.zoom = { enabled: true };
  }
  if (!isEmpty(slotMapToolbarConfig)) {
    slots.toolbar = { ...slots.toolbar, ...slotMapToolbarConfig };
  }
  return slots;
};
const iterateChartInteractionsSlots = (children) => {
  let interactions = {};
  Children.forEach(children, (childInteractions) => {
    if (isSlot(childInteractions, ChartInteractions.Zoom)) {
      interactions = { ...interactions, zoom: { enabled: true } };
    } else if (isSlot(childInteractions, ChartInteractions.ZoomToFit)) {
      interactions = { ...interactions, zoomToFit: { enabled: true } };
    }
  });
  return interactions;
};
const iterateChartToolbarSlots = (children) => {
  let toolbar = {};
  Children.forEach(children, (childToolbar) => {
    if (isSlot(childToolbar, ChartToolbar.DownloadData)) {
      toolbar = {
        ...toolbar,
        downloadCSV: { enabled: true }
      };
    }
  });
  return toolbar;
};
export {
  iterateConfigSlots
};
