// @dynatrace/strato-geo/esm/map/utils/tooltip-type-guards.js
import { isDefaultTooltipHandler } from "../hooks/use-tooltip-template.js";
import { BubbleLayerTooltip } from "../slots/BubbleLayerTooltip.js";
import { ChoroplethLayerTooltip } from "../slots/ChoroplethLayerTooltip.js";
import { ConnectionLayerTooltip } from "../slots/ConnectionLayerTooltip.js";
import { DotLayerTooltip } from "../slots/DotLayerTooltip.js";
const isDotTooltipTemplate = (template) => typeof template === typeof DotLayerTooltip;
const isBubbleTooltipTemplate = (template) => typeof template === typeof BubbleLayerTooltip;
const isConnectionTooltipTemplate = (template) => typeof template === typeof ConnectionLayerTooltip;
const isChoroplethTooltipTemplate = (template) => typeof template === typeof ChoroplethLayerTooltip;
const isConnection = (data) => "path" in data.data;
const isRegion = (data) => "name" in data;
export {
  isBubbleTooltipTemplate,
  isChoroplethTooltipTemplate,
  isConnection,
  isConnectionTooltipTemplate,
  isDefaultTooltipHandler,
  isDotTooltipTemplate,
  isRegion
};
