// @dynatrace/strato-components/tables/SimpleTable/hooks/useColumns.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useColumns_exports = {};
__export(useColumns_exports, {
  isSimpleTableGroupColumnDef: () => isSimpleTableGroupColumnDef,
  useColumns: () => useColumns
});
module.exports = __toCommonJS(useColumns_exports);
var import_react = require("react");
function isSimpleTableGroupColumnDef(columnDef) {
  return columnDef !== null && typeof columnDef === "object" && "columns" in columnDef && Array.isArray(columnDef.columns);
}
function isSimpleTableDisplayColumnDef(columnDef) {
  return columnDef !== null && typeof columnDef === "object" && "accessor" in columnDef;
}
function splitAccessor(column) {
  if (typeof column.accessor === "string") {
    return {
      accessorKey: column.accessor
    };
  }
  if (typeof column.accessor === "function") {
    return {
      accessorFn: column.accessor
    };
  }
  return {};
}
function createHeaderRenderer(header) {
  if (typeof header === "function") {
    return () => header();
  }
  return header;
}
function createDisplayColumnDef(columnDef) {
  const composedDef = {
    id: columnDef.id,
    header: createHeaderRenderer(columnDef.header),
    meta: {
      width: columnDef.width,
      minWidth: columnDef.minWidth,
      maxWidth: columnDef.maxWidth,
      alignment: columnDef.alignment ?? "left"
    },
    ...splitAccessor(columnDef)
  };
  if (columnDef.cell) {
    const customCell = columnDef.cell;
    if (typeof customCell === "string") {
      composedDef.cell = customCell;
      return composedDef;
    }
    if (typeof customCell === "function") {
      return {
        ...composedDef,
        cell: createCellRenderer(customCell)
      };
    }
    return composedDef;
  }
  return composedDef;
}
function createCellRenderer(consumerCustomCell) {
  return ({ getValue, row, column }) => {
    return consumerCustomCell({
      value: getValue(),
      rowIndex: row.index,
      rowData: row.original,
      rowId: row.id
    });
  };
}
function useColumns(propColumns) {
  return (0, import_react.useMemo)(() => {
    const createdColumns = [];
    for (const firstLevelColumn of propColumns) {
      if (isSimpleTableGroupColumnDef(firstLevelColumn)) {
        const groupColumn = {
          id: firstLevelColumn.id,
          header: createHeaderRenderer(firstLevelColumn.header),
          columns: firstLevelColumn.columns?.map(
            (childColumn) => createDisplayColumnDef(childColumn)
          ),
          meta: {
            width: firstLevelColumn.width,
            alignment: firstLevelColumn.alignment ?? "left"
          }
        };
        createdColumns.push(groupColumn);
      } else if (isSimpleTableDisplayColumnDef(firstLevelColumn)) {
        createdColumns.push(createDisplayColumnDef(firstLevelColumn));
      }
    }
    return createdColumns;
  }, [propColumns]);
}
