// @dynatrace/strato-components/charts/core/components/axes/components/CategoryBottomAxis.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CategoryBottomAxis_exports = {};
__export(CategoryBottomAxis_exports, {
  CategoryBottomAxis: () => CategoryBottomAxis
});
module.exports = __toCommonJS(CategoryBottomAxis_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_scale = require("@visx/scale");
var import_DivAxisWrapper = require("./DivAxisWrapper.js");
var import_AxisBottom = require("../AxisBottom.js");
var import_constants = require("../constants.js");
var import_TickLabel = require("./TickLabel.js");
var import_calculate_axis_size = require("../utils/calculate-axis-size.js");
var import_get_tick_label_props = require("../utils/get-tick-label-props.js");
var import_is_tick_label_visible = require("../utils/is-tick-label-visible.js");
var import_getCategoryBottomAxisProps = require("./utils/getCategoryBottomAxisProps.js");
const CategoryBottomAxis = (props) => {
  const {
    scale,
    containerWidth,
    categories,
    containerHeight,
    tickLabelLayout = "horizontal",
    label,
    truncationMode = "end"
  } = props;
  const ticks = (0, import_scale.getTicks)(scale);
  const tickNames = ticks.map((tickId) => categories[tickId]) ?? "Unknown";
  const isTickRotated = tickLabelLayout === "vertical";
  const isTickLabelVisible = (0, import_is_tick_label_visible.isTickLabelSpaceAvailable)(scale.step());
  const maxSize = (0, import_calculate_axis_size.getXTickMaxSize)(
    containerHeight,
    containerWidth,
    isTickRotated,
    scale,
    label
  );
  const tickLabelProps = (0, import_get_tick_label_props.getTickLabelPropsForCategoryAxis)(
    "bottom",
    tickLabelLayout
  );
  const tickNameHasEmoji = tickNames.some((name) => import_constants.EMOJI_REGEX.test(name));
  const tickFormat = (_, index) => tickNames[index];
  return isTickLabelVisible ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_DivAxisWrapper.DivAxisWrapper,
    {
      divAxisStyle: import_constants.axisFontProps,
      ticks: tickNames,
      label,
      isTickVertical: true,
      isTickRotated,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_AxisBottom.AxisBottom,
        {
          role: "graphics-symbol",
          "aria-label": "category axis bottom",
          scale,
          containerSize: containerWidth,
          height: containerHeight,
          tickComponent: ({
            x: renderX,
            y: renderY,
            ...rest
          }) => {
            const {
              x,
              y,
              foreignObjectInlineStyle,
              foreignObjectContentInlineStyle
            } = (0, import_getCategoryBottomAxisProps.getCategoryBottomAxisProps)(
              renderX,
              isTickRotated,
              maxSize,
              tickNameHasEmoji
            );
            return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_TickLabel.TickLabel,
              {
                foreignObjectInlineStyle,
                foreignObjectContentInlineStyle,
                isTickRotated,
                truncationMode,
                maxSize,
                x,
                y,
                ...rest
              }
            );
          },
          tickValues: ticks,
          tickLabelProps,
          tickFormat,
          label,
          width: containerWidth
        }
      )
    }
  ) : null;
};
CategoryBottomAxis["displayName"] = "CategoryBottomAxis";
