// @dynatrace/strato-components/charts/core/components/axes/components/CategoryLeftAxis.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CategoryLeftAxis_exports = {};
__export(CategoryLeftAxis_exports, {
  CategoryLeftAxis: () => CategoryLeftAxis
});
module.exports = __toCommonJS(CategoryLeftAxis_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_scale = require("@visx/scale");
var import_constants = require("../../../constants.js");
var import_AxisLeft = require("../AxisLeft.js");
var import_constants2 = require("../constants.js");
var import_DivAxisWrapper = require("./DivAxisWrapper.js");
var import_TickLabel = require("./TickLabel.js");
var import_calculate_axis_size = require("../utils/calculate-axis-size.js");
var import_get_tick_label_props = require("../utils/get-tick-label-props.js");
var import_is_tick_label_visible = require("../utils/is-tick-label-visible.js");
var import_getCategoryLeftAxisProps = require("./utils/getCategoryLeftAxisProps.js");
const TICK_LABEL_PROPS = (0, import_get_tick_label_props.getTickLabelPropsForCategoryAxis)("left");
const CategoryLeftAxis = (props) => {
  const {
    scale,
    containerWidth,
    containerHeight,
    label,
    categories,
    truncationMode = import_constants.DEFAULT_TRUNCATION_MODE,
    isVerticalLayout = true
  } = props;
  const ticks = (0, import_scale.getTicks)(scale);
  const tickNames = ticks.map((tickId) => categories[tickId]) ?? "Unknown";
  const maxSize = (0, import_calculate_axis_size.getYTickMaxSize)(containerWidth, label);
  const tickFormat = (_, index) => tickNames[index];
  const isTickLabelVisible = (0, import_is_tick_label_visible.isTickLabelSpaceAvailable)(scale.step());
  const tickNameHasEmoji = tickNames.some((name) => import_constants2.EMOJI_REGEX.test(name));
  return isTickLabelVisible ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_DivAxisWrapper.DivAxisWrapper,
    {
      divAxisStyle: import_constants2.axisFontProps,
      ticks: tickNames,
      label,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_AxisLeft.AxisLeft,
        {
          role: "graphics-symbol",
          "aria-label": "category axis left",
          scale,
          containerSize: containerHeight,
          width: containerWidth,
          tickComponent: ({
            x: renderX,
            y: renderY,
            ...rest
          }) => {
            const {
              x,
              y,
              foreignObjectInlineStyle,
              foreignObjectContentInlineStyle
            } = (0, import_getCategoryLeftAxisProps.getCategoryLeftAxisProps)(
              renderX,
              renderY,
              isVerticalLayout,
              maxSize,
              tickNameHasEmoji
            );
            return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_TickLabel.TickLabel,
              {
                foreignObjectInlineStyle,
                foreignObjectContentInlineStyle,
                truncationMode,
                maxSize,
                x,
                y,
                ...rest
              }
            );
          },
          tickValues: ticks,
          tickLabelProps: TICK_LABEL_PROPS,
          tickFormat,
          label
        }
      )
    }
  ) : null;
};
CategoryLeftAxis["displayName"] = "CategoryLeftAxis";
