// @dynatrace/strato-components/charts/core/components/axes/components/DivAxisWrapper.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DivAxisWrapper_exports = {};
__export(DivAxisWrapper_exports, {
  DivAxisWrapper: () => DivAxisWrapper
});
module.exports = __toCommonJS(DivAxisWrapper_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_Flex = require("../../../../../layouts/flex/Flex.js");
var import_constants = require("../constants.js");
const TICK_MIN_WIDTH = 8;
const TICK_MIN_HEIGHT = 12;
const getDivContainerAxisStyle = (isTickVertical, isTickRotated) => {
  let tickStyle;
  let labelStyle;
  const rotatedTextStyle = {
    writingMode: "vertical-rl"
  };
  if (isTickVertical) {
    tickStyle = {
      marginTop: import_constants.X_AXES_TICK_LINE_OFFSET,
      minHeight: isTickRotated ? TICK_MIN_HEIGHT : 0,
      whiteSpace: "nowrap",
      overflow: "hidden",
      ...isTickRotated ? rotatedTextStyle : {}
    };
    labelStyle = {
      marginTop: import_constants.X_AXES_LABEL_OFFSET,
      whiteSpace: "nowrap",
      overflow: "hidden"
    };
  } else {
    tickStyle = {
      marginRight: import_constants.Y_AXES_TICK_LINE_OFFSET,
      minWidth: TICK_MIN_WIDTH,
      whiteSpace: "nowrap"
    };
    labelStyle = {
      marginRight: import_constants.Y_AXIS_LABEL_OFFSET,
      minWidth: import_constants.AXIS_LABEL_HEIGHT,
      whiteSpace: "nowrap",
      ...rotatedTextStyle
    };
  }
  return { tickStyle, labelStyle };
};
const DivAxisWrapper = ({
  ticks,
  label,
  isTickVertical = false,
  isTickRotated = false,
  divAxisStyle = {},
  children
}) => {
  const { tickStyle, labelStyle } = getDivContainerAxisStyle(
    isTickVertical,
    isTickRotated
  );
  const containerAxisStyle = {
    position: "relative",
    /* A combination of pointer Events is needed here and in visibleSVGAxisStyle to allow the correct functioning with the resizer in the bottom legend when the tick labels are rotated and truncated (edge case) */
    pointerEvents: "none"
  };
  const visibleSvgAxisStyle = {
    position: "absolute",
    pointerEvents: "visible"
  };
  const hiddenDivAxisStyle = {
    visibility: "hidden",
    ...divAxisStyle
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { style: containerAxisStyle, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { style: visibleSvgAxisStyle, children }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      import_Flex.Flex,
      {
        style: hiddenDivAxisStyle,
        flexDirection: isTickVertical ? "column-reverse" : "row",
        gap: 0,
        children: [
          label && /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { style: labelStyle, children: label }),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Flex.Flex, { flexDirection: isTickVertical ? "row" : "column", children: ticks?.map((tick, index) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { style: tickStyle, children: tick }, `${index}_${tick}`)) })
        ]
      }
    )
  ] });
};
DivAxisWrapper["displayName"] = "DivAxisWrapper";
