// @dynatrace/strato-components/charts/core/components/axes/components/TickLabel.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TickLabel_exports = {};
__export(TickLabel_exports, {
  TickLabel: () => TickLabel
});
module.exports = __toCommonJS(TickLabel_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_Tooltip = require("../../../../..//overlays/tooltip/Tooltip.js");
var import_is_browser_safari = require("../../../utils/is-browser-safari.js");
var import_chars_font_width_map = require("../../text-ellipsis/chars-font-width-map.js");
var import_truncate_string = require("../../text-ellipsis/truncate-string.js");
const TickLabel = (props) => {
  const {
    foreignObjectInlineStyle,
    foreignObjectContentInlineStyle,
    formattedValue,
    maxSize,
    truncationMode,
    x,
    y
  } = props;
  if (!formattedValue) {
    return null;
  }
  const truncatedValue = (0, import_truncate_string.truncateString)(
    formattedValue,
    maxSize,
    import_chars_font_width_map.charsFontWidth12pixelsSize400Map,
    truncationMode
  );
  const disableTooltip = formattedValue === truncatedValue;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("g", { children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    "foreignObject",
    {
      x,
      y,
      height: foreignObjectInlineStyle.height,
      width: foreignObjectInlineStyle.width,
      style: { ...foreignObjectInlineStyle },
      children: [
        (0, import_is_browser_safari.isSafari)() && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          "div",
          {
            style: { ...foreignObjectContentInlineStyle },
            title: formattedValue,
            children: truncatedValue
          }
        ),
        !(0, import_is_browser_safari.isSafari)() && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_Tooltip.Tooltip,
          {
            disabled: disableTooltip,
            text: formattedValue,
            style: { display: "inline-block" },
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              "div",
              {
                style: {
                  ...foreignObjectContentInlineStyle,
                  position: "relative"
                },
                children: truncatedValue
              }
            )
          }
        )
      ]
    }
  ) });
};
