// @dynatrace/strato-components/charts/core/components/axes/components/TimeTableHeaderAxis.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TimeTableHeaderAxis_exports = {};
__export(TimeTableHeaderAxis_exports, {
  TimeTableHeaderAxis: () => TimeTableHeaderAxis
});
module.exports = __toCommonJS(TimeTableHeaderAxis_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_scale = require("@visx/scale");
var import_DivAxisWrapper = require("./DivAxisWrapper.js");
var import_format_date_ticks = require("../../../utils/format-date-ticks.js");
var import_format_value = require("../../../utils/format-value.js");
var import_ticks = require("../../../utils/ticks.js");
var import_AxisTableHeader = require("../AxisTableHeader.js");
var import_constants = require("../constants.js");
var import_get_tick_label_props = require("../utils/get-tick-label-props.js");
const TimeTableHeaderAxis = (props) => {
  const { scale, containerWidth, containerHeight, stroke, min, max } = props;
  const numTicks = (0, import_ticks.getGanttHorizontalTimeAxisNumTicks)(containerWidth);
  const ticks = (0, import_scale.getTicks)(scale, numTicks);
  const tickValues = (0, import_format_value.getTickValues)(
    ticks,
    import_format_date_ticks.formatDateTicks
  );
  const tickFormat = (_, index) => tickValues[index];
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_DivAxisWrapper.DivAxisWrapper,
    {
      divAxisStyle: import_constants.axisFontProps,
      ticks: tickValues,
      isTickVertical: true,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_AxisTableHeader.AxisTableHeader,
        {
          role: "graphics-symbol",
          "aria-label": "time axis tableheader",
          scale,
          containerSize: containerWidth,
          height: containerHeight,
          numTicks,
          tickValues: ticks,
          tickLabelProps: (0, import_get_tick_label_props.getTickLabelPropsForGanttAxis)(min, max),
          tickFormat,
          stroke
        }
      )
    }
  );
};
TimeTableHeaderAxis["displayName"] = "TimeTableHeaderAxis";
