// @dynatrace/strato-components/charts/core/components/axes/components/TimeBottomAxis.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TimeBottomAxis_exports = {};
__export(TimeBottomAxis_exports, {
  TimeBottomAxis: () => TimeBottomAxis
});
module.exports = __toCommonJS(TimeBottomAxis_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_AxisBottom = require("../AxisBottom.js");
var import_constants = require("../constants.js");
var import_DivAxisWrapper = require("./DivAxisWrapper.js");
var import_get_tick_label_props = require("../utils/get-tick-label-props.js");
const TimeBottomAxis = (props) => {
  const {
    scale,
    containerWidth,
    containerHeight,
    label,
    stroke,
    ticks,
    tickColor,
    formattedTicks,
    shrinkTickEdges
  } = props;
  const tickFormat = (_, index) => formattedTicks[index];
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_DivAxisWrapper.DivAxisWrapper,
    {
      divAxisStyle: import_constants.axisFontProps,
      ticks: formattedTicks,
      label,
      isTickVertical: true,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_AxisBottom.AxisBottom,
        {
          role: "graphics-symbol",
          "aria-label": "time axis bottom",
          scale,
          containerSize: containerWidth,
          height: containerHeight,
          tickValues: ticks,
          tickLabelProps: shrinkTickEdges ? import_get_tick_label_props.bottomPanTickLabelDateProps : import_get_tick_label_props.bottomTickLabelDateProps,
          tickFormat,
          tickColor,
          label,
          stroke
        }
      )
    }
  );
};
TimeBottomAxis["displayName"] = "TimeBottomAxis";
