// @dynatrace/strato-components/charts/core/components/axes/components/ValueLeftAxis.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ValueLeftAxis_exports = {};
__export(ValueLeftAxis_exports, {
  ValueLeftAxis: () => ValueLeftAxis
});
module.exports = __toCommonJS(ValueLeftAxis_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_DivAxisWrapper = require("./DivAxisWrapper.js");
var import_AxisLeft = require("../AxisLeft.js");
var import_constants = require("../constants.js");
var import_get_tick_label_props = require("../utils/get-tick-label-props.js");
const TICK_LABEL_PROPS = (0, import_get_tick_label_props.getTickLabelPropsForValueAxis)("left");
const ValueLeftAxis = (props) => {
  const {
    scale,
    containerHeight,
    containerWidth,
    label,
    stroke,
    ticks,
    formattedTicks
  } = props;
  const tickFormat = (_, index) => formattedTicks[index];
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_DivAxisWrapper.DivAxisWrapper,
    {
      divAxisStyle: import_constants.axisFontProps,
      ticks: formattedTicks,
      label,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_AxisLeft.AxisLeft,
        {
          role: "graphics-symbol",
          "aria-label": "value axis left",
          scale,
          containerSize: containerHeight,
          width: containerWidth,
          tickValues: ticks,
          tickLabelProps: TICK_LABEL_PROPS,
          tickFormat,
          label,
          stroke
        }
      )
    }
  );
};
ValueLeftAxis["displayName"] = "ValueLeftAxis";
