// @dynatrace/strato-components/charts/core/components/threshold-bar/indicators/ThresholdIndicator.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ThresholdIndicator_exports = {};
__export(ThresholdIndicator_exports, {
  ThresholdIndicator: () => ThresholdIndicator
});
module.exports = __toCommonJS(ThresholdIndicator_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_shape = require("@visx/shape");
var import_useId = require("../../../../../core/hooks/useId.js");
var import_ThresholdBrightnessFilter = require("../ThresholdBrightnessFilter.js");
const ThresholdIndicator = (props) => {
  const id = (0, import_useId.useId)();
  const { data, color, isActive = false, strokeOnly = false } = props;
  if (data.length === 0) {
    return null;
  }
  const opacity = isActive ? 0.2 : 0.1;
  const strokeWidth = isActive || strokeOnly ? 1 : 0;
  const strokeDasharray = !isActive && strokeOnly ? "3,2" : "0,0";
  const lineProps = {
    data,
    stroke: color,
    strokeWidth,
    strokeDasharray
  };
  const brightnessFilterId = `threshold-indicator-${id}`;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ThresholdBrightnessFilter.ThresholdBrightnessFilter, { id: brightnessFilterId }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_shape.LinePath,
      {
        ...lineProps,
        role: "graphics-symbol",
        "aria-label": "threshold indicator line 0",
        x: (d) => d.x0,
        y: (d) => d.y0
      }
    ),
    !strokeOnly && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_shape.Area,
      {
        data,
        role: "graphics-symbol",
        "aria-label": "threshold indicator area",
        x0: (d) => d.x0,
        x1: (d) => d.x1,
        y0: (d) => d.y0,
        y1: (d) => d.y1,
        fill: color,
        opacity,
        filter: isActive ? `url(#${brightnessFilterId})` : void 0
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_shape.LinePath,
      {
        ...lineProps,
        role: "graphics-symbol",
        "aria-label": "threshold indicator line 1",
        x: (d) => d.x1,
        y: (d) => d.y1
      }
    )
  ] });
};
