// @dynatrace/strato-components/charts/core/slots/toolbar/CompactToolbar.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CompactToolbar_exports = {};
__export(CompactToolbar_exports, {
  CompactToolbar: () => CompactToolbar,
  baseToolbarProps: () => baseToolbarProps,
  sanitizeProps: () => sanitizeProps
});
module.exports = __toCommonJS(CompactToolbar_exports);
var import_lodash_es = require("lodash");
const baseToolbarProps = () => {
  return {
    placement: "top-right",
    hidden: false,
    "data-dt-content": "",
    "data-dt-name": "",
    "data-dt-children-name": "",
    "data-dt-features": "",
    "data-dt-properties": "",
    "data-dt-mouse-over": "",
    "data-dt-value-content": false,
    "data-dt-children-value-content": false,
    "data-dt-ancestor": ""
  };
};
const sanitizeProps = (externalProps, allowedProps = ["placement", "hidden"]) => {
  return (0, import_lodash_es.pick)(externalProps, allowedProps);
};
const CompactToolbar = (prefix) => {
  const component = (props) => null;
  component["displayName"] = `${prefix}.Toolbar`;
  return component;
};
