// @dynatrace/strato-components/esm/charts/xy-chart-adoption/timeseries/TimeseriesToXY.js
import { jsx, jsxs } from "react/jsx-runtime";
import { forwardRef } from "react";
import { iterateSeriesSlots } from "./iterate-slots.js";
import { transformTimeseriesDataToXYChart } from "./transformer-data-to-xy.js";
import {
  DEFAULT_VARIANT
} from "./types/timeseries.js";
import {
  defineAxisId,
  renderAreaSeries,
  renderBandSeries,
  renderBarSeries,
  renderLineSeries,
  xAxisId
} from "./utils/render-series.js";
import { XYChart } from "../../xy-chart/XYChartRoot.js";
const TimeseriesToXY = forwardRef((props, _ref) => {
  const { data = [], variant = DEFAULT_VARIANT, children } = props;
  const { line, bar, area, band, shapePriority } = iterateSeriesSlots(children);
  const buckets = {
    line: variant === "line" ? [...data, ...line] : line,
    area: variant === "area" ? [...data, ...area] : area,
    bar: variant === "bar" ? [...data, ...bar] : bar,
    band
  };
  const order = [...shapePriority];
  if (data.length > 0 && !order.includes(variant)) {
    order.push(variant);
  }
  const result = transformTimeseriesDataToXYChart(buckets, order);
  const seriesForShape = (shape) => {
    switch (shape) {
      case "line":
        return result.line.map(renderLineSeries);
      case "area":
        return result.area.map(renderAreaSeries);
      case "bar":
        return result.bar.map(renderBarSeries);
      case "band":
        return result.band.map(renderBandSeries);
    }
  };
  return /* @__PURE__ */ jsxs(XYChart, { data: [], children: [
    order.flatMap(seriesForShape),
    /* @__PURE__ */ jsx(
      XYChart.XAxis,
      {
        id: xAxisId,
        type: "time",
        position: "bottom",
        min: "data-min",
        max: "data-max"
      }
    ),
    [...result.units].map((unit, index) => /* @__PURE__ */ jsx(
      XYChart.YAxis,
      {
        id: defineAxisId(unit),
        type: "numerical",
        unit,
        position: index % 2 === 0 ? "left" : "right"
      },
      defineAxisId(unit)
    ))
  ] });
});
export {
  TimeseriesToXY
};
