// @dynatrace/strato-components/esm/charts/xy-chart-adoption/timeseries/transformer-data-to-xy.js
import { singleLineName } from "../../core/utils/single-line-name.js";
const mapTimeseries = (series, seriesId) => {
  const { datapoints, name, unit, ...extraSeriesProps } = series;
  const seriesName = singleLineName(name);
  const data = datapoints.map((datapoint) => ({
    ...extraSeriesProps,
    start: datapoint.start,
    end: datapoint.end,
    center: datapoint.center,
    value: datapoint.value,
    seriesName,
    seriesId
  }));
  return { data, unit, seriesId };
};
const mapBand = (band, seriesId) => {
  const { datapoints, name, unit, ...extraSeriesProps } = band;
  const seriesName = singleLineName(name);
  const data = datapoints.map((datapoint) => ({
    ...extraSeriesProps,
    start: datapoint.start,
    end: datapoint.end,
    y0: datapoint.y0,
    y1: datapoint.y1,
    seriesName,
    seriesId
  }));
  return { data, unit, seriesId };
};
function transformTimeseriesDataToXYChart(input, shapeRenderOrder) {
  const units = /* @__PURE__ */ new Set();
  const result = {
    line: [],
    area: [],
    bar: [],
    band: [],
    units
  };
  let globalIndex = 0;
  shapeRenderOrder.forEach((shape) => {
    if (shape === "band") {
      input.band.forEach((band) => {
        const mapped = mapBand(band, `${shape}-${globalIndex}`);
        result.band.push(mapped);
        units.add(mapped.unit);
        globalIndex++;
      });
    } else {
      input[shape].forEach((series) => {
        const mapped = mapTimeseries(series, `${shape}-${globalIndex}`);
        result[shape].push(mapped);
        units.add(mapped.unit);
        globalIndex++;
      });
    }
  });
  return result;
}
export {
  transformTimeseriesDataToXYChart
};
