// @dynatrace/strato-components/esm/tables/DataTable/components/Toolbar/DataTableToolbar.js
import { jsx, jsxs } from "react/jsx-runtime";
import {
  Children,
  forwardRef
} from "react";
import { DataTableColumnOrderSettings } from "./ColumnSettings/DataTableColumnOrderSettings.js";
import { DataTableColumnPinningSettings } from "./ColumnSettings/DataTableColumnPinningSettings.js";
import { DataTableColumnSettings } from "./ColumnSettings/DataTableColumnSettings.js";
import { DataTableColumnSettingsTrigger } from "./ColumnSettings/DataTableColumnSettingsTrigger.js";
import { DataTableColumnVisibilitySettings } from "./ColumnSettings/DataTableColumnVisibilitySettings.js";
import { DataTableToolbarMenu } from "./DataTableToolbarMenu.js";
import { useBehavioralTrackingProps } from "../../../../core/hooks/useBehavioralTrackingProps.js";
import {
  Intent
} from "../../../../core/slots/Intent/Intent.js";
import { isSlot } from "../../../../core/utils/is-slot.js";
import { devLog } from "../../../../core/utils/logging.js";
import { Flex } from "../../../../layouts/flex/Flex.js";
import { DataTableDownload } from "../../features/Download/Download.js";
import {
  DataTableGlobalLineWrapAction,
  DataTableGlobalLineWrapToolbarAction
} from "../../features/LineWrap/LineWrap.js";
function _isDataTableToolbar(child) {
  return child?.type === DataTableToolbar;
}
function getToolbarConfig(children) {
  const slots = {
    visibilitySettings: false,
    columnOrderSettings: false,
    columnPinningSettings: false,
    downloadData: false,
    globalLineWrap: false
  };
  const toolbarItemProps = { intents: [] };
  Children.forEach(children, (child) => {
    if (isSlot(child, DataTableColumnVisibilitySettings)) {
      slots.visibilitySettings = true;
      toolbarItemProps.resetColumnVisibility = child.props?.resetColumnVisibility;
    } else if (isSlot(child, DataTableColumnOrderSettings)) {
      slots.columnOrderSettings = true;
      toolbarItemProps.resetColumnOrder = child.props?.resetColumnOrder;
    } else if (isSlot(child, DataTableColumnPinningSettings)) {
      slots.columnPinningSettings = true;
      toolbarItemProps.resetColumnPinning = child.props?.resetColumnPinning;
    } else if (isSlot(child, DataTableColumnSettingsTrigger)) {
      toolbarItemProps.customTriggerProps = child.props;
    } else if (isSlot(child, DataTableDownload)) {
      slots.downloadData = true;
      toolbarItemProps.downloadExcludedColumns = child.props?.excludeColumns;
      toolbarItemProps.onDownloadData = child.props?.onDownloadData;
    } else if (isSlot(child, DataTableGlobalLineWrapAction)) {
      slots.globalLineWrap = true;
    } else if (isSlot(child, Intent)) {
      toolbarItemProps.intents.push(child.props);
    } else {
      devLog("Invalid child of DataTableToolbar");
    }
  });
  return { slots, toolbarItemProps };
}
const DataTableToolbar = forwardRef(
  (props, forwardedRef) => {
    const {
      children,
      "data-testid": dataTestId,
      className: consumerClassName,
      style: consumerStyle,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
    const { slots, toolbarItemProps } = getToolbarConfig(children);
    if (Children.count(children) === 0) {
      return null;
    }
    return /* @__PURE__ */ jsxs(
      Flex,
      {
        ref: forwardedRef,
        justifyContent: "flex-end",
        alignItems: "flex-end",
        gap: 4,
        "data-testid": dataTestId,
        className: consumerClassName,
        style: {
          gridArea: "1 / 2 / 2 / 2",
          ...consumerStyle
        },
        ...behavioralTrackingProps,
        ...restProps,
        children: [
          (slots.visibilitySettings || slots.columnOrderSettings || slots.columnPinningSettings) && /* @__PURE__ */ jsx(
            DataTableColumnSettings,
            {
              customTriggerProps: toolbarItemProps?.customTriggerProps
            }
          ),
          slots.globalLineWrap && /* @__PURE__ */ jsx(DataTableGlobalLineWrapToolbarAction, {}),
          (slots.downloadData || toolbarItemProps.intents.length > 0) && /* @__PURE__ */ jsx(
            DataTableToolbarMenu,
            {
              hasDownloadData: slots.downloadData,
              downloadExcludedColumns: toolbarItemProps?.downloadExcludedColumns,
              intents: toolbarItemProps.intents
            }
          )
        ]
      }
    );
  }
);
DataTableToolbar.displayName = "DataTable.Toolbar";
export {
  DataTableToolbar,
  _isDataTableToolbar,
  getToolbarConfig
};
