// @dynatrace/strato-components/esm/tables/DataTable/components/Toolbar/DataTableToolbarMenu.js
import { jsx, jsxs } from "react/jsx-runtime";
import { Fragment, useRef, useState } from "react";
import { useIntl } from "react-intl";
import { DotMenuIcon, DownloadIcon } from "@dynatrace/strato-icons";
import { DataTableDownloadData } from "./Download/DataTableDownloadData.js";
import { Button } from "../../../../buttons/button/Button.js";
import { Menu } from "../../../../navigation/menu/Menu.js";
import { Tooltip } from "../../../../overlays/tooltip/Tooltip.js";
import { dataTableMessages } from "../../messages.js";
const DataTableToolbarMenu = (props) => {
  const { hasDownloadData, downloadExcludedColumns, intents } = props;
  const [menuOpen, setMenuOpen] = useState(false);
  const [isDownloading, setIsDownloading] = useState(false);
  const hasIntents = intents.length > 0;
  const triggerIcon = hasIntents ? /* @__PURE__ */ jsx(DotMenuIcon, {}) : /* @__PURE__ */ jsx(DownloadIcon, {});
  const Group = hasIntents ? Menu.Group : Fragment;
  const intl = useIntl();
  let toolbarMenuMessage = dataTableMessages.downloadButtonLabel;
  if (isDownloading) {
    toolbarMenuMessage = dataTableMessages.downloadingButtonLabel;
  } else if (hasIntents) {
    toolbarMenuMessage = dataTableMessages.toolbarButtonLabel;
  }
  const toolbarMenuLabel = intl.formatMessage(toolbarMenuMessage);
  const simpleButtonRef = useRef(null);
  if (!hasDownloadData && !hasIntents) {
    return;
  }
  if (menuOpen === false) {
    return /* @__PURE__ */ jsx(Tooltip, { text: toolbarMenuLabel, children: /* @__PURE__ */ jsx(
      Button,
      {
        ref: simpleButtonRef,
        variant: "default",
        "aria-label": toolbarMenuLabel,
        onClick: () => setMenuOpen(true),
        loading: isDownloading,
        children: /* @__PURE__ */ jsx(Button.Prefix, { children: triggerIcon })
      }
    ) });
  }
  const {
    intentsWithAppOrIntentId: intentsWithAppId,
    intentsWithoutAppOrIntentId: intentsWithoutAppId
  } = intents.reduce(
    (acc, intentProps, idx) => {
      const hasAppOrIntentId = intentProps.options.recommendedAppId !== void 0 || intentProps.options.recommendedIntentId !== void 0;
      const intent = /* @__PURE__ */ jsxs(Menu.Intent, { ...intentProps, children: [
        intentProps.icon && /* @__PURE__ */ jsx(Menu.Prefix, { children: intentProps.icon }),
        intentProps.children
      ] }, `intent-${idx}`);
      hasAppOrIntentId ? acc.intentsWithAppOrIntentId.push(intent) : acc.intentsWithoutAppOrIntentId.push(intent);
      return acc;
    },
    { intentsWithAppOrIntentId: [], intentsWithoutAppOrIntentId: [] }
  );
  return /* @__PURE__ */ jsxs(
    Menu,
    {
      open: menuOpen,
      onOpenChange: (open) => {
        if (open === false) {
          requestAnimationFrame(() => {
            simpleButtonRef.current?.focus();
          });
        }
        setMenuOpen(open);
      },
      modal: false,
      children: [
        /* @__PURE__ */ jsx(Tooltip, { text: toolbarMenuLabel, children: /* @__PURE__ */ jsx(Menu.Trigger, { children: /* @__PURE__ */ jsx(Button, { variant: "default", "aria-label": toolbarMenuLabel, children: /* @__PURE__ */ jsx(Button.Prefix, { children: triggerIcon }) }) }) }),
        /* @__PURE__ */ jsxs(Menu.Content, { alignment: "end", children: [
          hasDownloadData && /* @__PURE__ */ jsx(Group, { children: /* @__PURE__ */ jsx(
            DataTableDownloadData,
            {
              excludeColumns: downloadExcludedColumns,
              hasIntents,
              isDownloading,
              onDownloadStart: () => setIsDownloading(true),
              onDownloadEnd: () => setIsDownloading(false)
            }
          ) }),
          intentsWithAppId.length > 0 && /* @__PURE__ */ jsx(Group, { children: intentsWithAppId }),
          intentsWithoutAppId.length > 0 && /* @__PURE__ */ jsx(Group, { children: intentsWithoutAppId })
        ] })
      ]
    }
  );
};
export {
  DataTableToolbarMenu
};
