// @dynatrace/strato-components/esm/charts/core/utils/color-contrast.js
const DARK_LABEL_TEXT = "#000000";
const LIGHT_LABEL_TEXT = "#ffffff";
function sRGBtoLinear(sRGB) {
  const sRGB_Threshold = 0.04045;
  return sRGB <= sRGB_Threshold ? sRGB / 12.92 : Math.pow((sRGB + 0.055) / 1.055, 2.4);
}
function getLuminance(rgbColor) {
  const { r, g, b } = rgbColor;
  const RsRGB = r / 255;
  const GsRGB = g / 255;
  const BsRGB = b / 255;
  const luminance = 0.2126 * sRGBtoLinear(RsRGB) + 0.7152 * sRGBtoLinear(GsRGB) + 0.0722 * sRGBtoLinear(BsRGB);
  return luminance;
}
function getContrastRatio(foreground, background) {
  const fgLum = getLuminance(foreground);
  const bgLum = getLuminance(background);
  return fgLum >= bgLum ? (fgLum + 0.05) / (bgLum + 0.05) : (bgLum + 0.05) / (fgLum + 0.05);
}
const getColorOptions = () => ({
  dark: { r: 0, g: 0, b: 0 },
  light: { r: 255, g: 255, b: 255 }
});
const getRGBfromHex = (hex) => {
  const r = parseInt(hex.slice(1, 3), 16);
  const g = parseInt(hex.slice(3, 5), 16);
  const b = parseInt(hex.slice(5, 7), 16);
  return { r, g, b };
};
function getColorTokenDefaultValue(token) {
  const firstComma = token.indexOf(",");
  return token.substring(firstComma + 1, token.length - 1).trim();
}
function getColorByContrast(bgColorHex, lightColor = LIGHT_LABEL_TEXT, darkColor = DARK_LABEL_TEXT) {
  const { light, dark } = getColorOptions();
  const bgColorRGB = getRGBfromHex(bgColorHex);
  const lightRatio = getContrastRatio(light, bgColorRGB);
  const darkRatio = getContrastRatio(dark, bgColorRGB);
  return lightRatio >= darkRatio ? lightColor : darkColor;
}
export {
  DARK_LABEL_TEXT,
  LIGHT_LABEL_TEXT,
  getColorByContrast,
  getColorTokenDefaultValue,
  getContrastRatio
};
