// @dynatrace/strato-components/esm/charts/core/utils/core-scale.js
import { scaleLinear } from "d3-scale";
import { isNumber, isUndefined } from "lodash-es";
import { getSymlogScale } from "./symlog-scale.js";
function scaleBandInvert(scale) {
  const domain = scale.domain();
  const paddingOuter = scale(domain[0]) ?? 0;
  const numberOfBands = scale.step();
  return function(mousePosition) {
    const halfInnerPadding = scale.step() * scale.paddingInner() / 2;
    const index = Math.floor(
      (mousePosition + halfInnerPadding - paddingOuter) / numberOfBands
    );
    return domain[Math.max(0, Math.min(index, domain.length - 1))];
  };
}
const buildScale = (scaleType, range, dataBoundaries, configMinMax) => {
  const minMaxOptions = {
    niceMin: isDefaultBoundary(configMinMax?.min),
    niceMax: isDefaultBoundary(configMinMax?.max)
  };
  const { min, max } = getMinMax(
    scaleType,
    dataBoundaries,
    configMinMax,
    minMaxOptions
  );
  return getScale(scaleType, [min, max], range);
};
const getScale = (scaleType, domain, range = [0, 1]) => (scaleType === "linear" ? scaleLinear() : getSymlogScale()).domain(domain).range(range);
const MIN_MAX_DEFAULT_OPTIONS = {
  niceMin: true,
  niceMax: true
};
const getMinMax = (scaleType, dataBoundaries, configMinMax, options = MIN_MAX_DEFAULT_OPTIONS) => {
  const baselineDomain = getBaselineDomain(scaleType, dataBoundaries);
  let min = getScaleBoundary(
    dataBoundaries.min,
    baselineDomain.min,
    configMinMax?.min
  );
  let max = getScaleBoundary(
    dataBoundaries.max,
    baselineDomain.max,
    configMinMax?.max
  );
  const niceScale = getScale(scaleType, [min, max]).nice().domain();
  const scale = getScale(scaleType, [min, max]).domain();
  const [defaultMin, defaultMax] = scale;
  const [niceMin, niceMax] = niceScale;
  min = defaultMin;
  max = defaultMax;
  if (options.niceMin) {
    min = niceMin;
  }
  if (options.niceMax) {
    max = niceMax;
  }
  return { min, max };
};
const getBaselineDomain = (scaleType, dataBoundaries) => {
  let [baselineMin, baselineMax] = [0, 0];
  if (scaleType === "log") {
    baselineMin = 1;
    baselineMax = -1;
  }
  const baselineDomainMin = dataBoundaries.min > baselineMin ? baselineMin : dataBoundaries.min;
  const baselineDomainMax = dataBoundaries.max < baselineMax ? baselineMax : dataBoundaries.max;
  return { min: baselineDomainMin, max: baselineDomainMax };
};
const getScaleBoundary = (dataBoundary, baselineBoundary, scaleBoundary) => {
  if (isNumber(scaleBoundary)) {
    return scaleBoundary;
  }
  if (isDefaultBoundary(scaleBoundary)) {
    return baselineBoundary;
  }
  return dataBoundary;
};
const isDefaultBoundary = (boundary) => isUndefined(boundary) || boundary === "auto";
const isDataBoundary = (boundary) => !isUndefined(boundary) && (boundary === "data-min" || boundary === "data-max");
export {
  buildScale,
  getBaselineDomain,
  getMinMax,
  getScale,
  getScaleBoundary,
  isDataBoundary,
  isDefaultBoundary,
  scaleBandInvert
};
