// @dynatrace/strato-components/esm/charts/core/utils/format-value.js
import { isEmpty, uniq } from "lodash-es";
import { format } from "@dynatrace-sdk/units";
import { formatDateTicks } from "./format-date-ticks.js";
import { getFormatterDefinition } from "./format-value-by-type.js";
const DEFAULT_PRECISION = 1;
const MAX_PRECISION_ALLOWED = 20;
const formatValue = (value, unit = "", precision = DEFAULT_PRECISION) => {
  if (typeof unit === "string") {
    return format(value, {
      ...getFormatterDefinition(unit),
      maximumFractionDigits: precision
    });
  }
  return format(value, { input: unit, maximumFractionDigits: precision });
};
function getFormatterPrecision(tickValues, formatter, unit) {
  for (let minPrecision = DEFAULT_PRECISION; minPrecision <= MAX_PRECISION_ALLOWED; minPrecision++) {
    const formattedTicks = tickValues.map(
      (value) => formatter(value, unit, minPrecision)
    );
    if (allUniq(formattedTicks)) {
      return minPrecision;
    }
  }
  return MAX_PRECISION_ALLOWED;
}
const allUniq = (arr) => uniq(arr).length === arr.length;
const defaultTickValueFormatter = (value, unit) => {
  if (typeof value === "string") {
    return `${value}${isEmpty(unit) ? "" : ` ${unit}`}`;
  }
  if (value instanceof Date) {
    return formatDateTicks(value);
  }
  return formatValue(value, unit, DEFAULT_PRECISION);
};
const tickValueFormatter = (tickFormatter, unit) => (value) => tickFormatter && typeof value === "number" ? tickFormatter(value, unit) : defaultTickValueFormatter(value, unit);
const getTickValues = (originalTicks, formatter) => originalTicks.map((value) => formatter?.(value) ?? value.toString());
export {
  DEFAULT_PRECISION,
  MAX_PRECISION_ALLOWED,
  defaultTickValueFormatter,
  formatValue,
  getFormatterPrecision,
  getTickValues,
  tickValueFormatter
};
