// @dynatrace/strato-components/esm/charts/core/utils/get-data-point-is-gap.js
function getDatapointIsGap(datapoint, nextDatapoint, threshold = 0) {
  return hasGap(datapoint, nextDatapoint, threshold);
}
function hasGap(currentDatapoint, nextDatapoint, threshold) {
  if (!currentDatapoint || !nextDatapoint) {
    return false;
  }
  const distance = Math.abs(getDistance(nextDatapoint, currentDatapoint));
  const hasGap2 = distance > 0;
  const isGapOutsideThreshold = threshold < distance;
  return hasGap2 && isGapOutsideThreshold;
}
function isNumericalDatapoint(datapoint) {
  return typeof datapoint.start === "number";
}
function isTimeframeDatapoint(datapoint) {
  return datapoint.start instanceof Date;
}
function getDistance(nextDatapoint, currentDatapoint) {
  if (isNumericalDatapoint(nextDatapoint) && isNumericalDatapoint(currentDatapoint)) {
    return nextDatapoint.start - currentDatapoint.start;
  }
  if (isTimeframeDatapoint(nextDatapoint) && isTimeframeDatapoint(currentDatapoint)) {
    return nextDatapoint.start.getTime() - resolveEndDate(currentDatapoint).getTime();
  }
  return 0;
}
function resolveEndDate(datapoint) {
  return datapoint.end ?? datapoint.start;
}
export {
  getDatapointIsGap
};
