// @dynatrace/strato-components/esm/charts/core/utils/text-truncation.js
const truncateText = (text, truncationMode, stopTruncation) => {
  let truncatedText = text;
  while (!stopTruncation(truncatedText)) {
    const filteredText = truncatedText.replace(/\.\.\./, "");
    if (!filteredText.length) {
      break;
    }
    truncatedText = TRUNCATE_LABEL[truncationMode](filteredText);
  }
  return truncatedText;
};
const ELLIPSIS = "...";
const middleTruncation = (text) => {
  const middle = Math.floor(text.length / 2);
  const firstHalf = text.slice(0, middle);
  const secondHalf = text.slice(middle + 1);
  const hasEllipsis = firstHalf.length || secondHalf.length;
  const newText = `${firstHalf}${hasEllipsis ? ELLIPSIS : ""}${secondHalf}`;
  return newText;
};
const startTruncation = (text) => {
  const slicedText = text.slice(1);
  const hasEllipsis = slicedText.length;
  const newText = `${hasEllipsis ? ELLIPSIS : ""}${slicedText}`;
  return newText;
};
const endTruncation = (text) => {
  const slicedText = text.slice(0, -1);
  const hasEllipsis = slicedText.length;
  const newText = `${text.slice(0, -1)}${hasEllipsis ? ELLIPSIS : ""}`;
  return newText;
};
const TRUNCATE_LABEL = {
  start: startTruncation,
  end: endTruncation,
  middle: middleTruncation
};
export {
  ELLIPSIS,
  TRUNCATE_LABEL,
  endTruncation,
  middleTruncation,
  startTruncation,
  truncateText
};
