// @dynatrace/strato-components/esm/charts/core/utils/validators.js
const isSameType = (a, b) => {
  return typeof a === typeof b;
};
const VALUE_MATCHERS = {
  equals: (subject, value) => `${subject}` === `${value}`,
  "not-equals": (subject, value) => `${subject}` !== `${value}`,
  "less-than": (subject, value) => isSameType(subject, value) && subject < value,
  "less-or-equal": (subject, value) => isSameType(subject, value) && subject <= value,
  "greater-than": (subject, value) => isSameType(subject, value) && subject > value,
  "greater-or-equal": (subject, value) => isSameType(subject, value) && subject >= value,
  "starts-with": (subject, value) => `${subject}`.startsWith(`${value}`),
  "not-starts-with": (subject, value) => !`${subject}`.startsWith(`${value}`),
  "ends-with": (subject, value) => `${subject}`.endsWith(`${value}`),
  "not-ends-with": (subject, value) => !`${subject}`.endsWith(`${value}`),
  contains: (subject, value) => `${subject}`.includes(`${value}`),
  "not-contains": (subject, value) => !`${subject}`.includes(`${value}`),
  "matches-phrase": (subject, value) => subject === value,
  "not-matches-phrase": (subject, value) => subject !== value
};
const RANGE_MATCHERS = {
  "less-than": (subject, value) => isSameType(subject.to, value) && subject.to < value,
  "less-or-equal": (subject, value) => isSameType(subject.to, value) && subject.to <= value,
  "greater-than": (subject, value) => isSameType(subject.to, value) && subject.from > value,
  "greater-or-equal": (subject, value) => isSameType(subject.to, value) && subject.from >= value,
  equals: (subject, value) => subject.to === value && subject.from === value,
  "not-equals": (subject, value) => !(subject.to === value && subject.from === value)
};
const isRangeOperator = (operator) => {
  return operator in RANGE_MATCHERS;
};
function buildValidator(matchers) {
  return (subject, comparator, value) => {
    if (comparator in matchers) {
      return matchers[comparator](
        subject,
        value
      );
    }
    return false;
  };
}
export {
  RANGE_MATCHERS,
  VALUE_MATCHERS,
  buildValidator,
  isRangeOperator
};
