// @dynatrace/strato-components/esm/charts/core/utils/zoom.js
import { inRange } from "lodash-es";
import { GRAIL_MAX_DOMAIN, GRAIL_MIN_DOMAIN } from "../constants.js";
const PAN_LARGE_STEP = 0.02;
const PAN_SMALL_STEP = 0.01;
function isDomainMinZoom(newDomain, infiniteZoom, initialDomain) {
  if (newDomain[0] <= +GRAIL_MIN_DOMAIN && newDomain[1] >= +GRAIL_MAX_DOMAIN) {
    return true;
  }
  if (initialDomain === void 0 || infiniteZoom) {
    return false;
  }
  if (newDomain[0] === initialDomain[0] && newDomain[1] !== initialDomain[1] || newDomain[0] !== initialDomain[0] && newDomain[1] === initialDomain[1]) {
    return false;
  }
  return !(inRange(newDomain[0], initialDomain[0], initialDomain[1]) && inRange(newDomain[1], initialDomain[0], initialDomain[1]));
}
function isDomainMaxZoom(newDomain, initialDomain) {
  const newMinValue = newDomain[0];
  const newMaxValue = newDomain[1];
  const isEqualDomain = newMinValue === newMaxValue;
  if (isEqualDomain) {
    return true;
  }
  return initialDomain[0] > initialDomain[1] ? newMaxValue > newMinValue : newMinValue > newMaxValue;
}
function isZoomBoundaries(initialDomain, nextDomain, infiniteZoom) {
  const isMinZoom = isDomainMinZoom(nextDomain, infiniteZoom, initialDomain);
  const isMaxZoom = isDomainMaxZoom(nextDomain, initialDomain);
  return { isMinZoom, isMaxZoom };
}
const handlePanningMovement = (xScale, panDirection, stepType, initialXScaleDomain) => {
  const [, width] = xScale.range();
  const directionFactor = panDirection === "left" ? 1 : -1;
  const panStep = (stepType === "small" ? PAN_SMALL_STEP : PAN_LARGE_STEP) * width;
  return panningMove(xScale, initialXScaleDomain, directionFactor * panStep);
};
const panningMove = (xScale, initialXScaleDomain, delta) => {
  const [rangeMin, rangeMax] = xScale.range();
  const [initialMin, initialMax] = initialXScaleDomain ?? [];
  const newMin = xScale.invert(rangeMin - delta);
  const newMax = xScale.invert(rangeMax - delta);
  if (+newMin > +initialMin && +newMax < +initialMax || !initialXScaleDomain) {
    return [newMin, newMax];
  }
  return [xScale.invert(rangeMin), xScale.invert(rangeMax)];
};
function getA11yToolbarConfig(config) {
  const { zoom, zoomX, download, intent } = config;
  const a11YToolbar = [
    ...[{ explore: 0 }],
    ...zoom?.enabled ? [{ "zoom-in": -1 }] : [],
    ...zoom?.enabled ? [{ "zoom-out": -1 }] : [],
    ...zoom?.enabled || zoomX?.enabled ? [{ reset: -1 }] : [],
    ...download?.enabled ? [{ download: -1 }] : [],
    ...intent ? [{ intent: -1 }] : [],
    ...[{ collapse: -1 }]
  ];
  return a11YToolbar;
}
function isSameDomain(currentDomain, initialDomain) {
  if (currentDomain === void 0) {
    return true;
  }
  return +currentDomain[0] === +initialDomain[0] && +currentDomain[1] === +initialDomain[1];
}
export {
  getA11yToolbarConfig,
  handlePanningMovement,
  isDomainMaxZoom,
  isDomainMinZoom,
  isSameDomain,
  isZoomBoundaries
};
