// @dynatrace/strato-components/esm/core/utils/accessors/parse-accessor.js
var TokenType = /* @__PURE__ */ ((TokenType2) => {
  TokenType2[TokenType2["DOUBLE_QUOTE"] = 0] = "DOUBLE_QUOTE";
  TokenType2[TokenType2["DOT"] = 1] = "DOT";
  TokenType2[TokenType2["OTHER"] = 2] = "OTHER";
  return TokenType2;
})(TokenType || {});
const ESCAPE_CHARACTER = "\\";
function getAccessorParts(accessor) {
  let currentPart = "";
  let openedQuote = false;
  const accessorParts = [];
  for (const token of tokenize(accessor)) {
    if (token.type === 0 /* DOUBLE_QUOTE */) {
      openedQuote = !openedQuote;
    } else if (openedQuote) {
      currentPart += token.value;
    } else if (token.type === 1 /* DOT */) {
      accessorParts.push(currentPart);
      currentPart = "";
    } else {
      currentPart += token.value;
    }
  }
  if (currentPart.length > 0) {
    accessorParts.push(currentPart);
  }
  return accessorParts;
}
function* tokenize(value) {
  let escaping = false;
  for (const chr of value) {
    if (escaping) {
      escaping = false;
      yield buildToken(chr, true);
    } else if (chr === ESCAPE_CHARACTER) {
      escaping = true;
    } else {
      yield buildToken(chr);
    }
  }
  if (escaping) {
    throw new Error(`Unexpected escape sequence at index ${value.length - 1}`);
  }
}
function getAxisRootAccessor(accessor) {
  const parts = getAccessorParts(accessor);
  return parts[0];
}
function getAxisExtensionAccessor(accessor) {
  const parts = getAccessorParts(accessor);
  return parts[1];
}
const buildToken = (chr, escaped = false) => {
  const tokenType = escaped ? 2 /* OTHER */ : getTokenType(chr);
  return {
    type: tokenType,
    value: chr
  };
};
const getTokenType = (chr) => {
  switch (chr) {
    case ".":
      return 1 /* DOT */;
    case '"':
      return 0 /* DOUBLE_QUOTE */;
    default:
      return 2 /* OTHER */;
  }
};
export {
  TokenType,
  getAccessorParts,
  getAxisExtensionAccessor,
  getAxisRootAccessor,
  tokenize
};
