// @dynatrace/strato-components/esm/layouts/input-group/InputGroup.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import { Children, forwardRef } from "react";
import * as styles from "./InputGroup.sty.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { Grid } from "../grid/Grid.js";
const COMPONENT_NAME = "InputGroup";
const InputGroup = /* @__PURE__ */ forwardRef((inputGroupProps, ref) => {
  const {
    children,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    // @ts-expect-error - We want to allow overriding data-dt-component as we use the component internally in other Strato components, but it should not appear in the public API.
    "data-dt-component": dataDtComponent = COMPONENT_NAME,
    ...remainingProps
  } = inputGroupProps;
  const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
  if (Children.count(children) === 0) {
    return null;
  }
  return /* @__PURE__ */ jsx(
    Grid,
    {
      ref,
      gridAutoFlow: "column",
      gridAutoRows: "max-content",
      justifyContent: "start",
      columnGap: 2,
      rowGap: 4,
      className: clsx(consumerClassName, styles.inputGroup),
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...restProps,
      ...behavioralTrackingProps,
      "data-dt-component": dataDtComponent,
      children
    }
  );
});
InputGroup.displayName = COMPONENT_NAME;
export {
  InputGroup
};
