// @dynatrace/strato-components/esm/tables/DataTable/features/RowSelection/RowSelectionCells.js
import { jsx } from "react/jsx-runtime";
import {
  useEffect
} from "react";
import { useIntl } from "react-intl";
import { Checkbox } from "../../../../forms/checkbox/Checkbox.js";
import { Tooltip } from "../../../../overlays/tooltip/Tooltip.js";
import { DataTableDefaultCell } from "../../components/DataTableDefaultCell.js";
import { dataTableMessages } from "../../messages.js";
function RowSelectionHeader(props) {
  const { table } = props;
  const resetBatchSelection = table.resetBatchSelection;
  const value = getCheckboxValue(table);
  const isDisabled = isSelectAllDisabled(table);
  const intl = useIntl();
  const checkboxMessage = intl.formatMessage(
    dataTableMessages.selectAllTooltip,
    {
      nextSelection: value === false || value === "indeterminate" ? "select" : "deselect",
      scope: table.options.rowSelectionSelectAllMode
    }
  );
  useEffect(() => {
    document.addEventListener("click", resetBatchSelection);
    return () => {
      document.removeEventListener("click", resetBatchSelection);
    };
  }, [resetBatchSelection]);
  function toggleSelectAll(event) {
    if (isDisabled) {
      event?.preventDefault();
      return;
    }
    if (table.options.rowSelectionSelectAllMode === "page") {
      if (table.options.rowSelectionLimit !== void 0) {
        table.getToggleAllLimitedPageRowsSelectedHandler()(event);
      } else {
        table.getToggleAllPageRowsSelectedHandler()(event);
      }
    } else if (table.options.rowSelectionSelectAllMode === "all") {
      if (table.options.rowSelectionLimit !== void 0) {
        table.getToggleAllLimitedRowsSelectedHandler()(event);
      } else {
        table.getToggleAllRowsSelectedHandler()(event);
      }
    }
  }
  const selectAllDtProperties = `state:${value === true ? "deselect" : "select"}`;
  const checkbox = /* @__PURE__ */ jsx(
    Checkbox,
    {
      value: isDisabled ? false : value,
      "aria-disabled": isDisabled,
      "aria-label": checkboxMessage,
      onChange: (_, event) => toggleSelectAll(event),
      "data-dt-name": "RowSelectDeselectAll",
      "data-dt-properties": selectAllDtProperties
    }
  );
  return /* @__PURE__ */ jsx(DataTableDefaultCell, { children: isDisabled ? checkbox : /* @__PURE__ */ jsx(Tooltip, { text: checkboxMessage, children: checkbox }) });
}
function isSelectAllDisabled(table) {
  const isTableEmpty = table.getRowCount() === 0;
  if (isTableEmpty) {
    return true;
  }
  let isAllRowsSelectionDisabled = false;
  if (table.options.rowSelectionSelectAllMode === "page") {
    isAllRowsSelectionDisabled = table.getIsAllPageRowsSelectionDisabled();
  } else if (table.options.rowSelectionSelectAllMode === "all") {
    isAllRowsSelectionDisabled = table.getIsAllRowsSelectionDisabled();
  }
  return isAllRowsSelectionDisabled;
}
function getCheckboxValue(table) {
  if (table.options.rowSelectionSelectAllMode === "page") {
    if (table.getIsAllPageRowsSelected()) {
      return true;
    }
    if (table.getIsSomePageRowsSelected()) {
      return "indeterminate";
    }
    return false;
  }
  if (table.options.rowSelectionSelectAllMode === "all") {
    if (table.getIsAllRowsSelected()) {
      return true;
    }
    if (table.getIsSomeRowsSelected()) {
      return "indeterminate";
    }
    return false;
  }
  return false;
}
const resolveCheckboxValue = (row) => {
  if (row.subRows.length === 0) {
    return row.getIsSelected();
  }
  if (row.getIsAllSubRowsSelected()) {
    return true;
  }
  if (row.getIsSomeSelected()) {
    return "indeterminate";
  }
  return false;
};
function RowSelectionCell(props) {
  const { row } = props;
  const intl = useIntl();
  const isDisabled = row.getShouldDisableRowSelection();
  const rowDtProperties = `state:${resolveCheckboxValue(row) === true ? "deselect" : "select"}`;
  return /* @__PURE__ */ jsx(
    DataTableDefaultCell,
    {
      onMouseDown: (event) => {
        document.getSelection()?.removeAllRanges();
        event.preventDefault();
      },
      children: /* @__PURE__ */ jsx(
        Checkbox,
        {
          "aria-label": intl.formatMessage(dataTableMessages.selectRowCheckbox, {
            id: row.id
          }),
          value: resolveCheckboxValue(row),
          "aria-disabled": isDisabled,
          onChange: (value, event) => {
            if (isDisabled) {
              event?.preventDefault();
              return;
            }
            row.batchSelect(value, (event?.nativeEvent).shiftKey);
          },
          "data-row-selection-checkbox": true,
          "data-dt-name": "RowSelectDeselect",
          "data-dt-properties": rowDtProperties
        }
      )
    }
  );
}
export {
  RowSelectionCell,
  RowSelectionHeader
};
