// @dynatrace/strato-geo/map/hooks/use-geo-tooltip-sync.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var use_geo_tooltip_sync_exports = {};
__export(use_geo_tooltip_sync_exports, {
  useGeoOverlayTooltipState: () => useGeoOverlayTooltipState,
  useGeoTooltipSync: () => useGeoTooltipSync,
  useOverlayTooltipSelector: () => useOverlayTooltipSelector
});
module.exports = __toCommonJS(use_geo_tooltip_sync_exports);
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_charts = require("@dynatrace/strato-components/charts");
var import_build_geo_tooltip_state = require("../utils/build-geo-tooltip-state.js");
const useOverlayTooltipSelector = (selector) => {
  const store = (0, import_charts._useOverlayTooltipStore)();
  const prevRef = (0, import_react.useRef)(void 0);
  const getSnapshot = (0, import_react.useCallback)(() => {
    const next = selector(store.getState());
    if (prevRef.current !== void 0 && (0, import_lodash_es.isEqual)(prevRef.current, next)) {
      return prevRef.current;
    }
    prevRef.current = next;
    return next;
  }, [store, selector]);
  return (0, import_react.useSyncExternalStore)(store.subscribe, getSnapshot);
};
const useGeoOverlayTooltipState = () => {
  const metadata = useOverlayTooltipSelector(
    (state) => state.data?.metadata
  );
  const pinned = useOverlayTooltipSelector((state) => state.pinned);
  const inBounds = useOverlayTooltipSelector((state) => state.inBounds);
  return { data: metadata, pinned, inBounds };
};
const useGeoTooltipSync = (position, pinned, geometryType, geoData) => {
  const store = (0, import_charts._useOverlayTooltipStore)();
  const dispatch = (0, import_charts._useOverlayTooltipReducer)();
  (0, import_react.useEffect)(() => {
    const onScroll = () => {
      if (store.getState().pinned) {
        dispatch({ type: "TOOLTIP_MOUSE_LEAVE" })();
      }
    };
    document.body.addEventListener("scroll", onScroll, true);
    return () => {
      document.body.removeEventListener("scroll", onScroll, true);
    };
  }, [store, dispatch]);
  (0, import_react.useEffect)(() => {
    if (position) {
      const nextState = (0, import_build_geo_tooltip_state.buildGeoTooltipState)({
        absoluteX: position.x,
        absoluteY: position.y,
        relativeX: position.x,
        relativeY: position.y,
        pinned,
        geometry: geometryType,
        metadata: geoData
      });
      const current = store.getState();
      if (current.absoluteX !== nextState.absoluteX || current.absoluteY !== nextState.absoluteY || current.inBounds !== nextState.inBounds || current.pinned !== nextState.pinned || !(0, import_lodash_es.isEqual)(current.data?.metadata, nextState.data.metadata)) {
        store.setState(nextState);
      }
    } else {
      const currentState = store.getState();
      if (currentState.inBounds || currentState.pinned) {
        dispatch({ type: "SET_OUT_BOUNDS" })();
      }
    }
  }, [
    position?.x,
    position?.y,
    pinned,
    geometryType,
    geoData,
    store,
    dispatch
  ]);
};
