// @dynatrace/strato-geo/map/hooks/use-overlay-events.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var use_overlay_events_exports = {};
__export(use_overlay_events_exports, {
  useOverlayEvents: () => useOverlayEvents
});
module.exports = __toCommonJS(use_overlay_events_exports);
var import_charts = require("@dynatrace/strato-components/charts");
var import_store = require("../store/store.js");
var import_build_geo_tooltip_state = require("../utils/build-geo-tooltip-state.js");
var import_parse_tooltip_data = require("../utils/parse-tooltip-data.js");
const layerIdToGeometry = (layerId) => {
  if (layerId?.includes("bubble")) {
    return "geoBubble";
  }
  if (layerId?.includes("choropleth")) {
    return "geoChoropleth";
  }
  if (layerId?.includes("connection")) {
    return "geoConnection";
  }
  return "geoDot";
};
const useOverlayEvents = () => {
  const setOverlayState = (0, import_store.useSetStateOverlay)();
  const setState = (0, import_store.useSetState)();
  const dispatch = (0, import_charts._useOverlayTooltipReducer)();
  const store = (0, import_charts._useOverlayTooltipStore)();
  const overlay = (0, import_charts._useOverlayChart)();
  const getAbsolutePosition = (event) => {
    return {
      x: Math.round(event.originalEvent.clientX),
      y: Math.round(event.originalEvent.clientY)
    };
  };
  const setTooltipMarker = (hoveredLayerId, location) => {
    setState((prev) => ({
      ...prev,
      tooltip: {
        ...prev.tooltip,
        hoveredLayerId,
        location
      }
    }));
  };
  const hideTooltip = () => {
    dispatch({ type: "RESET_TOOLTIP" })();
    setTooltipMarker(void 0, void 0);
  };
  const handleMouseEnter = () => {
    setOverlayState({ mouseInBounds: true });
  };
  const handleMouseMove = (event) => {
    if (event.features?.length) {
      const currentState = store.getState();
      if (currentState.pinned) {
        return;
      }
      const { data, hoveredLayerId } = (0, import_parse_tooltip_data.extractDataFromEvent)(event);
      overlay.clear();
      setTooltipMarker(hoveredLayerId, event.lngLat);
      const pos = getAbsolutePosition(event);
      store.setState(
        (0, import_build_geo_tooltip_state.buildGeoTooltipState)({
          absoluteX: pos.x,
          absoluteY: pos.y,
          relativeX: Math.round(event.point.x),
          relativeY: Math.round(event.point.y),
          pinned: false,
          geometry: layerIdToGeometry(hoveredLayerId),
          metadata: data
        })
      );
    } else {
      const currentState = store.getState();
      if (!currentState.pinned) {
        hideTooltip();
      }
    }
  };
  const handleMouseLeave = () => {
    setOverlayState({ mouseInBounds: false });
    const currentState = store.getState();
    if (!currentState.pinned) {
      hideTooltip();
    }
  };
  const handleMouseClick = (event) => {
    const { data, featureId, hoveredLayerId } = (0, import_parse_tooltip_data.extractDataFromEvent)(event);
    if (!featureId) {
      hideTooltip();
      return;
    }
    if (!hoveredLayerId) {
      hideTooltip();
      return;
    }
    overlay.clear();
    setTooltipMarker(hoveredLayerId, event.lngLat);
    const pos = getAbsolutePosition(event);
    store.setState(
      (0, import_build_geo_tooltip_state.buildGeoTooltipState)({
        absoluteX: pos.x,
        absoluteY: pos.y,
        relativeX: Math.round(event.point.x),
        relativeY: Math.round(event.point.y),
        pinned: true,
        geometry: layerIdToGeometry(hoveredLayerId),
        metadata: data
      })
    );
  };
  const hideTooltipIfNotPinned = () => {
    const currentState = store.getState();
    if (!currentState.pinned) {
      hideTooltip();
    }
  };
  return {
    handleMouseEnter,
    handleMouseMove,
    handleMouseLeave,
    handleMouseClick,
    handleZoom: hideTooltipIfNotPinned,
    handleDrag: hideTooltipIfNotPinned
  };
};
