// @dynatrace/strato-components/core/utils/accessors/parse-accessor.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var parse_accessor_exports = {};
__export(parse_accessor_exports, {
  TokenType: () => TokenType,
  getAccessorParts: () => getAccessorParts,
  getAxisExtensionAccessor: () => getAxisExtensionAccessor,
  getAxisRootAccessor: () => getAxisRootAccessor,
  tokenize: () => tokenize
});
module.exports = __toCommonJS(parse_accessor_exports);
var TokenType = /* @__PURE__ */ ((TokenType2) => {
  TokenType2[TokenType2["DOUBLE_QUOTE"] = 0] = "DOUBLE_QUOTE";
  TokenType2[TokenType2["DOT"] = 1] = "DOT";
  TokenType2[TokenType2["OTHER"] = 2] = "OTHER";
  return TokenType2;
})(TokenType || {});
const ESCAPE_CHARACTER = "\\";
function getAccessorParts(accessor) {
  let currentPart = "";
  let openedQuote = false;
  const accessorParts = [];
  for (const token of tokenize(accessor)) {
    if (token.type === 0 /* DOUBLE_QUOTE */) {
      openedQuote = !openedQuote;
    } else if (openedQuote) {
      currentPart += token.value;
    } else if (token.type === 1 /* DOT */) {
      accessorParts.push(currentPart);
      currentPart = "";
    } else {
      currentPart += token.value;
    }
  }
  if (currentPart.length > 0) {
    accessorParts.push(currentPart);
  }
  return accessorParts;
}
function* tokenize(value) {
  let escaping = false;
  for (const chr of value) {
    if (escaping) {
      escaping = false;
      yield buildToken(chr, true);
    } else if (chr === ESCAPE_CHARACTER) {
      escaping = true;
    } else {
      yield buildToken(chr);
    }
  }
  if (escaping) {
    throw new Error(`Unexpected escape sequence at index ${value.length - 1}`);
  }
}
function getAxisRootAccessor(accessor) {
  const parts = getAccessorParts(accessor);
  return parts[0];
}
function getAxisExtensionAccessor(accessor) {
  const parts = getAccessorParts(accessor);
  return parts[1];
}
const buildToken = (chr, escaped = false) => {
  const tokenType = escaped ? 2 /* OTHER */ : getTokenType(chr);
  return {
    type: tokenType,
    value: chr
  };
};
const getTokenType = (chr) => {
  switch (chr) {
    case ".":
      return 1 /* DOT */;
    case '"':
      return 0 /* DOUBLE_QUOTE */;
    default:
      return 2 /* OTHER */;
  }
};
