// @dynatrace/strato-components/esm/charts/core/components/chart-layout/utils/get-areas-size-on-drag.js
import { calculateGraphSize } from "./calculate-graph-size.js";
import { ORIENTATION } from "../../../constants.js";
import {
  MIN_LEGEND_HEIGHT,
  MIN_LEGEND_WIDTH,
  MIN_PLOT_SIZE
} from "../types/constants.js";
const getAreasWidthSize = (containerHeight, containerWidth, minWidth, maxWidth, resizerWrapperSize, draggedAmount, legendContainer) => {
  const minLegendWidth = Math.min(containerWidth * minWidth, MIN_LEGEND_WIDTH);
  if (legendContainer.clientWidth - draggedAmount <= minLegendWidth) {
    const legend = {
      height: containerHeight,
      width: minLegendWidth
    };
    const graph = {
      width: calculateGraphSize(
        containerWidth,
        minLegendWidth,
        resizerWrapperSize
      ),
      height: containerHeight
    };
    return { legend, graph };
  }
  const availableWidth = containerWidth - resizerWrapperSize;
  const maxWidthAllowed = Math.max(
    availableWidth * maxWidth,
    availableWidth - MIN_PLOT_SIZE
  );
  if (legendContainer.clientWidth - draggedAmount >= maxWidthAllowed) {
    const legend = {
      height: containerHeight,
      width: maxWidthAllowed
    };
    const graph = {
      width: calculateGraphSize(
        containerWidth,
        maxWidthAllowed,
        resizerWrapperSize
      ),
      height: containerHeight
    };
    return { legend, graph };
  }
};
const getAreasHeightSize = (containerHeight, containerWidth, minHeight, maxHeight, resizerWrapperSize, draggedAmount, legendContainer) => {
  const minLegendHeight = Math.min(
    containerHeight * minHeight,
    MIN_LEGEND_HEIGHT
  );
  if (legendContainer.clientHeight - draggedAmount <= minLegendHeight) {
    const legend = {
      height: minLegendHeight,
      width: containerWidth
    };
    const graph = {
      height: calculateGraphSize(
        containerHeight,
        minLegendHeight,
        resizerWrapperSize
      ),
      width: containerWidth
    };
    return { legend, graph };
  }
  const availableHeight = containerHeight - resizerWrapperSize;
  const maxHeightAllowed = Math.max(
    availableHeight * maxHeight,
    availableHeight - MIN_PLOT_SIZE
  );
  if (legendContainer.clientHeight - draggedAmount >= maxHeightAllowed) {
    const legend = {
      height: maxHeightAllowed,
      width: containerWidth
    };
    const graph = {
      height: calculateGraphSize(
        containerHeight,
        maxHeightAllowed,
        resizerWrapperSize
      ),
      width: containerWidth
    };
    return { legend, graph };
  }
};
const getDefaultAreasSize = (areasSizes, draggedAmount, containerSize, resizerWrapperSize, orientation) => {
  const { graph, legend } = areasSizes;
  const isVertical = orientation === ORIENTATION.Vertical;
  const legendHeight = isVertical ? legend.height - draggedAmount : legend.height;
  const legendWidth = isVertical ? legend.width : legend.width - draggedAmount;
  const graphHeight = isVertical ? calculateGraphSize(containerSize, legendHeight, resizerWrapperSize) : graph.height;
  const graphWidth = isVertical ? graph.width : calculateGraphSize(containerSize, legendWidth, resizerWrapperSize);
  return {
    legend: { height: legendHeight, width: legendWidth },
    graph: { height: graphHeight, width: graphWidth }
  };
};
const getAreasSizeOnDragHandlers = (containerHeight, containerWidth, legendContainer, resizerWrapperSize) => ({
  getAreasHeightSizeOnDrag: (minHeight, maxHeight, verticalDraggedAmount = 0) => getAreasHeightSize(
    containerHeight,
    containerWidth,
    minHeight,
    maxHeight,
    resizerWrapperSize,
    verticalDraggedAmount,
    legendContainer
  ),
  getAreasWidthSizeOnDrag: (minWidth, maxWidth, horizontalDraggedAmount = 0) => getAreasWidthSize(
    containerHeight,
    containerWidth,
    minWidth,
    maxWidth,
    resizerWrapperSize,
    horizontalDraggedAmount,
    legendContainer
  )
});
export {
  getAreasSizeOnDragHandlers,
  getDefaultAreasSize
};
