// @dynatrace/strato-components/esm/charts/core/components/chart-layout/utils/get-chart-layout-settings.js
import { isUndefined } from "lodash-es";
import { extractLayoutSlots } from "./extract-layout-slots.js";
import { getOrientation } from "./orientation.js";
import { sanitizeLegendRatio } from "./sanitize-legend-ratio.js";
import { defaultRatio, ORIENTATION } from "../../../constants.js";
import {
  DEFAULT_LEGEND_CONFIG,
  DEFAULT_RATIO_BOUNDARIES
} from "../types/constants.js";
const getChartLayoutSettings = (containerSizes, children, isEmpty) => {
  const { provider, graphArea, legendArea, legendProps } = extractLayoutSlots(children);
  const containerAvailableArea = {
    width: containerSizes.width,
    height: containerSizes.height
  };
  if (!legendArea) {
    return {
      graphArea,
      containerAvailableArea,
      legendSettings: {
        orientation: ORIENTATION.Vertical,
        showResizer: false,
        hidden: true,
        ratio: 0,
        ratioBoundaries: DEFAULT_RATIO_BOUNDARIES
      },
      provider
    };
  }
  const {
    position,
    // eslint-disable-next-line @typescript-eslint/no-deprecated
    ratio = defaultRatio,
    initialRatio,
    onResize,
    overrideRatio,
    ratioBoundaries = DEFAULT_RATIO_BOUNDARIES
  } = legendProps ?? DEFAULT_LEGEND_CONFIG;
  const orientation = getOrientation(
    containerAvailableArea.width,
    containerAvailableArea.height,
    position
  );
  return {
    graphArea,
    legendArea,
    containerAvailableArea,
    legendSettings: {
      orientation,
      showResizer: !isEmpty,
      hidden: isUndefined(legendArea),
      ratio: sanitizeLegendRatio(ratio, ratioBoundaries, orientation),
      initialRatio: initialRatio ? sanitizeLegendRatio(initialRatio, ratioBoundaries, orientation) : void 0,
      overrideRatio,
      ratioBoundaries,
      onResize
    },
    provider
  };
};
export {
  getChartLayoutSettings
};
