// @dynatrace/strato-components/charts/core/components/canvas/utils/batched-path-context.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var batched_path_context_exports = {};
__export(batched_path_context_exports, {
  ARC_BATCH_SIZE: () => ARC_BATCH_SIZE,
  PATH_BATCH_SIZE: () => PATH_BATCH_SIZE,
  createBatchedStrokeContext: () => createBatchedStrokeContext
});
module.exports = __toCommonJS(batched_path_context_exports);
const PATH_BATCH_SIZE = 1e3;
const ARC_BATCH_SIZE = 1e3;
function createBatchedStrokeContext(ctx, batchSize) {
  let closePathCount = 0;
  const d3PathContext = {
    moveTo(x, y) {
      ctx.moveTo(x, y);
    },
    lineTo(x, y) {
      ctx.lineTo(x, y);
    },
    bezierCurveTo(cp1x, cp1y, cp2x, cp2y, x, y) {
      ctx.bezierCurveTo(cp1x, cp1y, cp2x, cp2y, x, y);
    },
    closePath() {
      ctx.closePath();
      closePathCount++;
      if (closePathCount >= batchSize) {
        ctx.stroke();
        ctx.beginPath();
        closePathCount = 0;
      }
    }
  };
  return d3PathContext;
}
