// @dynatrace/strato-components/charts/core/components/annotations/components/displays/AnnotationsNumericalDisplay.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var AnnotationsNumericalDisplay_exports = {};
__export(AnnotationsNumericalDisplay_exports, {
  AnnotationsNumericalDisplay: () => AnnotationsNumericalDisplay
});
module.exports = __toCommonJS(AnnotationsNumericalDisplay_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_AnnotationsContainer = require("./AnnotationsContainer.js");
var import_useBehavioralTrackingProps = require("../../../../../../core/hooks/useBehavioralTrackingProps.js");
var import_OverlayContext = require("../../../overlay-tooltip/contexts/OverlayContext.js");
var import_TooltipContext = require("../../../overlay-tooltip/contexts/TooltipContext.js");
var import_AnnotationsProviders = require("../../providers/AnnotationsProviders.js");
var import_AnnotationsRenderer = require("../AnnotationsRenderer.js");
var import_AnnotationsOverlayTooltip = require("../tooltip/AnnotationsOverlayTooltip.js");
var import_AnnotationsTooltipFormatter_context = require("../tooltip/contexts/AnnotationsTooltipFormatter.context.js");
const AnnotationsNumericalDisplay = (props) => {
  const {
    annotationsActions,
    markerGroupHover,
    formatter,
    tooltipConfig,
    "data-dt-component": dataDtComponent,
    ...rendererProps
  } = props;
  const [behaviorTrackingProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(props);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayContext.OverlayProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_AnnotationsContainer.AnnotationsContainer,
    {
      ...behaviorTrackingProps,
      "data-dt-component": dataDtComponent,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TooltipContext.TooltipProvider, { persistedStore: true, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_AnnotationsProviders.AnnotationsProviders,
        {
          onAnnotationsMarkersInteraction: markerGroupHover,
          annotationsActions,
          tooltipConfig,
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_AnnotationsTooltipFormatter_context.AnnotationsTooltipFormatterContext.Provider, { value: formatter, children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_AnnotationsRenderer.AnnotationsRenderer, { ...rendererProps }),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_AnnotationsOverlayTooltip.AnnotationsOverlayTooltip,
              {
                tooltipType: "numerical",
                truncationMode: rendererProps.truncateMode,
                maxLines: rendererProps.maxLines
              }
            )
          ] })
        }
      ) })
    }
  ) });
};
AnnotationsNumericalDisplay["displayName"] = "AnnotationsNumericalDisplay";
