// @dynatrace/strato-components/esm/content/markdown/transformers/table-mapping.js
import { Fragment, jsx } from "react/jsx-runtime";
import { isValidElement } from "react";
import Spacings from "@dynatrace/strato-design-tokens/spacings";
import { SimpleTable } from "../../../tables/SimpleTable/SimpleTable.js";
import { getChildren } from "../utils.js";
function getTextAlignment(cell) {
  return isValidElement(cell) ? cell?.props?.style?.textAlign : void 0;
}
const tableMappings = {
  table: ({ children, ...props }) => {
    if (children.length === 0) {
      return null;
    }
    const tHead = children[0];
    const tHeadRow = getChildren(tHead)[0];
    const columns = getChildren(
      tHeadRow
    ).map((cell, index) => ({
      header: () => /* @__PURE__ */ jsx(Fragment, { children: getChildren(cell) }),
      // eslint-disable-next-line react/jsx-no-useless-fragment
      cell: ({ value }) => /* @__PURE__ */ jsx(Fragment, { children: value }),
      accessor: String(index),
      id: String(index),
      alignment: getTextAlignment(cell)
    }));
    const tBody = children[1];
    const data = getChildren(tBody).map((row) => ({
      ...getChildren(row).map((cell) => getChildren(cell))
    }));
    return /* @__PURE__ */ jsx("div", { ...props, children: /* @__PURE__ */ jsx(
      SimpleTable,
      {
        style: {
          marginTop: Spacings.Size24,
          marginBottom: Spacings.Size24
        },
        columns,
        data,
        variant: {
          contained: true,
          verticalDividers: true,
          rowDensity: "default"
        }
      }
    ) });
  }
};
export {
  tableMappings
};
