// @dynatrace/strato-components/esm/content/markdown/transformers/typography-mapping.js
import { jsx } from "react/jsx-runtime";
import { CheckmarkSmallIcon } from "@dynatrace/strato-icons";
import {
  MarkdownCheckmarkBackgroundCSS,
  MarkdownHeadingCSS,
  MarkdownImageCSS,
  MarkdownListCSS,
  MarkdownListItemCSS,
  MarkdownParagraphCSS
} from "./typography-mapping.sty.js";
import { Divider } from "../../../layouts/divider/Divider.js";
import { Blockquote } from "../../../typography/block-quote/Blockquote.js";
import { Code } from "../../../typography/code/Code.js";
import { Emphasis } from "../../../typography/emphasis/Emphasis.js";
import { ExternalLink } from "../../../typography/external-link/ExternalLink.js";
import { Heading } from "../../../typography/heading/Heading.js";
import { List } from "../../../typography/list/List.js";
import { Paragraph } from "../../../typography/paragraph/Paragraph.js";
import { Strikethrough } from "../../../typography/strikethrough/Strikethrough.js";
import { Strong } from "../../../typography/strong/Strong.js";
import { Text } from "../../../typography/text/Text.js";
import { CodeSnippet } from "../../code-snippet/CodeSnippet.js";
import { matchLanguageParm } from "../utils.js";
const MarkdownCode = (props) => {
  const { className, inline, children } = props;
  const match = matchLanguageParm(className);
  let sanitizedChildren = children;
  if (Array.isArray(children)) {
    const firstChild = children[0];
    sanitizedChildren = [
      typeof firstChild === "string" ? firstChild.replace(/\n$/g, "") : ""
    ];
  } else if (typeof children === "string") {
    sanitizedChildren = children.replace(/\n$/g, "");
  }
  return inline ? /* @__PURE__ */ jsx(Code, { "data-testid": "code", children: sanitizedChildren }) : /* @__PURE__ */ jsx(
    CodeSnippet,
    {
      "data-testid": "code-snippet",
      language: match ? match[1] : void 0,
      children: sanitizedChildren
    }
  );
};
const typographyMappings = {
  code: MarkdownCode,
  // We need to add the inlineCode here since we use react-markdown in this repo
  // but mdx-js/react inside the cli-plugin
  inlineCode: ({ node, ...props }) => /* @__PURE__ */ jsx(Code, { ...props }),
  p: ({ children }) => /* @__PURE__ */ jsx(Paragraph, { className: MarkdownParagraphCSS, "data-testid": "paragraph", children }),
  strong: ({ ref, node, ...props }) => /* @__PURE__ */ jsx(Strong, { "data-testid": "strong", ...props }),
  a: ({ ref, href, node, ...props }) => /* @__PURE__ */ jsx(
    ExternalLink,
    {
      "data-testid": "external-link",
      href,
      ...props
    }
  ),
  input: ({ checked }) => /* @__PURE__ */ jsx(
    "span",
    {
      className: MarkdownCheckmarkBackgroundCSS,
      role: "checkbox",
      "aria-label": checked ? "checked" : "unchecked",
      "aria-checked": !!checked,
      "aria-disabled": true,
      children: checked ? /* @__PURE__ */ jsx(CheckmarkSmallIcon, {}) : null
    }
  ),
  ul: ({ children }) => /* @__PURE__ */ jsx(List, { className: MarkdownListCSS, children }),
  ol: ({ children }) => /* @__PURE__ */ jsx(List, { className: MarkdownListCSS, ordered: true, children }),
  li: ({ ref, node, ordered, ...props }) => {
    return /* @__PURE__ */ jsx(Text, { ...props, className: MarkdownListItemCSS });
  },
  h1: ({ ref, node, ...props }) => /* @__PURE__ */ jsx(
    Heading,
    {
      ...props,
      className: MarkdownHeadingCSS,
      "data-testid": "heading",
      level: 1,
      as: "h1"
    }
  ),
  h2: ({ ref, node, ...props }) => /* @__PURE__ */ jsx(
    Heading,
    {
      ...props,
      className: MarkdownHeadingCSS,
      "data-testid": "heading",
      level: 2,
      as: "h2"
    }
  ),
  h3: ({ ref, node, ...props }) => /* @__PURE__ */ jsx(
    Heading,
    {
      ...props,
      className: MarkdownHeadingCSS,
      "data-testid": "heading",
      level: 3,
      as: "h3"
    }
  ),
  h4: ({ ref, node, ...props }) => /* @__PURE__ */ jsx(
    Heading,
    {
      ...props,
      className: MarkdownHeadingCSS,
      "data-testid": "heading",
      level: 4,
      as: "h4"
    }
  ),
  h5: ({ ref, node, ...props }) => /* @__PURE__ */ jsx(
    Heading,
    {
      ...props,
      className: MarkdownHeadingCSS,
      "data-testid": "heading",
      level: 5,
      as: "h5"
    }
  ),
  h6: ({ ref, node, ...props }) => /* @__PURE__ */ jsx(
    Heading,
    {
      ...props,
      className: MarkdownHeadingCSS,
      "data-testid": "heading",
      level: 6,
      as: "h6"
    }
  ),
  del: ({ ref, node, ...props }) => /* @__PURE__ */ jsx(Strikethrough, { ...props }),
  em: ({ ref, node, ...props }) => /* @__PURE__ */ jsx(Emphasis, { ...props }),
  blockquote: ({ ref, node, ...props }) => /* @__PURE__ */ jsx(Blockquote, { ...props }),
  hr: () => /* @__PURE__ */ jsx(Divider, { orientation: "horizontal", as: "hr" }),
  img: ({ ref, alt, node, ...props }) => /* @__PURE__ */ jsx("img", { ...props, alt, className: MarkdownImageCSS })
};
export {
  typographyMappings
};
