// @dynatrace/strato-components/esm/filters/timeframe-selector/TimeframeSelector.js
import { jsx } from "react/jsx-runtime";
import { debounce } from "lodash-es";
import {
  forwardRef,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState
} from "react";
import { PresetItem } from "./components/presets/PresetItem.js";
import { Presets } from "./components/presets/Presets.js";
import { CustomTrigger } from "./components/trigger/CustomTrigger.js";
import { DisplayValue } from "./components/trigger/DisplayValue.js";
import { Trigger } from "./components/trigger/Trigger.js";
import { TIMEFRAMESELECTOR_ERROR_MESSAGES } from "./constants/timeframeSelectorMessages.js";
import { TimeframeSelectorRoot } from "./TimeframeSelectorRoot.js";
import { useCallbackRef } from "../../core/hooks/useCallbackRef.js";
import { useMergeRefs } from "../../core/hooks/useMergeRefs.js";
import { getTimeframeSelectorActions } from "../../core/time/timeframeStore/actions/timeframeSelector.js";
import { createTimeframeStore } from "../../core/time/timeframeStore/createTimeframeStore.js";
import { TimeframeSelectorStoreProvider } from "../../core/time/timeframeStore/timeframeStoreProvider.js";
import { useDateTimeMessages } from "../../core/time/useDateTimeMessages.js";
import { useMinMax, useMinMaxStoreSync } from "../../core/time/useMinMax.js";
import { validate } from "../../core/time/validation.js";
import { getTimeOffsetFromInitialValue } from "../../forms/calendar/utils/helpers.js";
import { useFormFieldMessagesContext } from "../../forms/forms-core/hooks/useFormFieldMessagesContext.js";
import { useSyncedFormControlProps } from "../../forms/forms-core/hooks/useSyncedFormControlProps.js";
import {
  GRAIL_MAX_DATE,
  GRAIL_MIN_DATE
} from "../../forms/forms-core/time/date-time.js";
import {
  removePrecisionFromTimeFrame,
  mapToPartialTimeframe
} from "../../forms/forms-core/time/timeParser.js";
const STEPPER_DEBOUNCE_MS = 300;
function useSyncPropsAndState(store, memoedValue, isControlled, precision, formatString, required, formFieldMessagesDispatch, timeframeSelectorRef, debouncedOnChangeRef) {
  useEffect(
    () => {
      isControlled && store.setState((state) => {
        debouncedOnChangeRef?.current?.cancel();
        const error = validate(memoedValue ?? null, state.min, state.max, {
          formMessageDispatch: formFieldMessagesDispatch,
          showError: false,
          checkOrder: true,
          ref: timeframeSelectorRef,
          expectedFormat: formatString,
          valueMode: "timeframeSelector",
          required,
          errorMessages: TIMEFRAMESELECTOR_ERROR_MESSAGES,
          formerError: state.timeframeSelectorError
        });
        const value = removePrecisionFromTimeFrame(
          memoedValue,
          state.precision
        );
        const timeOffset = getTimeOffsetFromInitialValue(
          value,
          "range",
          state.min,
          state.max,
          precision
        );
        return {
          ...state,
          timeframeSelectorTimeframe: value,
          calendarValue: value,
          dateTimeRangePickerTimeframe: value,
          timeframeSelectorError: error,
          calendarError: error,
          dateTimeRangePickerError: error,
          timeOffset,
          required
        };
      });
    },
    // timeframeSelectorRef and debouncedOnChangeRef are stable refs — no need to include in deps
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [
      formFieldMessagesDispatch,
      formatString,
      isControlled,
      memoedValue,
      precision,
      required,
      store
    ]
  );
  useEffect(() => {
    store.setState((state) => {
      const preciseValue = isControlled ? memoedValue : state.timeframeSelectorTimeframe;
      const value = removePrecisionFromTimeFrame(preciseValue, precision);
      return {
        ...state,
        precision,
        formatString,
        timeframeSelectorTimeframe: value,
        calendarValue: value,
        dateTimeRangePickerTimeframe: value
      };
    });
  }, [precision, formatString, store, isControlled, memoedValue]);
}
const _TimeframeSelector = forwardRef(
  ({
    value: propValue,
    defaultValue: propDefaultValue,
    onChange,
    min: propMin = GRAIL_MIN_DATE,
    max: propMax = GRAIL_MAX_DATE,
    disabled: propDisabled,
    required: propRequired,
    readOnly: propReadOnly,
    id: propId,
    "aria-disabled": propAriaDisabled,
    precision = "minutes",
    ...remainingProps
  }, forwardedRef) => {
    const [isControlled] = useState(propValue !== void 0);
    const onChangeProp = useCallbackRef(onChange);
    const memoedValue = useMemo(
      () => mapToPartialTimeframe(propValue, precision),
      [precision, propValue]
    );
    const memoedDefaultValue = useMemo(
      () => mapToPartialTimeframe(propDefaultValue, precision),
      /** We're ignoring updates to defaultValue as it only considers the initial state. */
      // eslint-disable-next-line react-hooks/exhaustive-deps
      []
    );
    const { min, max } = useMinMax(propMin, propMax);
    const calendarRef = useRef(null);
    const dateTimeFromRef = useRef(null);
    const dateTimeRangePickerRef = useRef(null);
    const timeframeSelectorRef = useRef(null);
    const { formatString } = useDateTimeMessages({ precision });
    const { dispatch: formFieldMessagesDispatch } = useFormFieldMessagesContext();
    const { disabled, required, id, ariaDisabled, readOnly } = useSyncedFormControlProps({
      propDisabled,
      propRequired,
      propId,
      propAriaDisabled,
      propReadOnly
    });
    const debouncedOnChangeRef = useRef(
      debounce((timeframe) => {
        onChangeProp?.(timeframe);
      }, STEPPER_DEBOUNCE_MS)
    );
    useEffect(() => {
      const debouncedOnChange = debouncedOnChangeRef.current;
      return () => debouncedOnChange.cancel();
    }, []);
    const [initialState] = useState(() => {
      const value = isControlled ? memoedValue : memoedDefaultValue;
      const error = validate(value, min, max, {
        formMessageDispatch: formFieldMessagesDispatch,
        precision,
        showError: false,
        checkOrder: true,
        ref: timeframeSelectorRef,
        expectedFormat: formatString,
        valueMode: "timeframeSelector",
        required,
        errorMessages: TIMEFRAMESELECTOR_ERROR_MESSAGES
      });
      const timeOffset = getTimeOffsetFromInitialValue(
        value,
        "range",
        min,
        max,
        precision
      );
      return {
        calendarValue: value,
        dateTimeRangePickerTimeframe: value,
        timeframeSelectorTimeframe: value,
        referenceTimeframe: null,
        min,
        max,
        precision,
        isFullDay: false,
        calendarRef,
        dateTimeFromRef,
        dateTimeRangePickerRef,
        timeframeSelectorRef,
        required,
        timeframeSelectorError: error,
        calendarError: error,
        dateTimeRangePickerError: error,
        previousTimeframe: null,
        timeOffset,
        formatString
      };
    });
    const [store] = useState(
      () => createTimeframeStore(
        initialState,
        getTimeframeSelectorActions(
          isControlled,
          timeframeSelectorRef,
          formFieldMessagesDispatch,
          onChangeProp,
          debouncedOnChangeRef,
          TIMEFRAMESELECTOR_ERROR_MESSAGES
        )
      )
    );
    useSyncPropsAndState(
      store,
      memoedValue,
      isControlled,
      precision,
      formatString,
      required,
      formFieldMessagesDispatch,
      timeframeSelectorRef,
      debouncedOnChangeRef
    );
    useMinMaxStoreSync(
      store,
      min,
      max,
      useCallback(
        (state, min2, max2) => {
          const error = validate(state.timeframeSelectorTimeframe, min2, max2, {
            formMessageDispatch: formFieldMessagesDispatch,
            formerError: state.timeframeSelectorError,
            precision: state.precision,
            showError: false,
            checkOrder: true,
            ref: timeframeSelectorRef,
            expectedFormat: state.formatString,
            valueMode: "timeframeSelector",
            required: state.required,
            errorMessages: TIMEFRAMESELECTOR_ERROR_MESSAGES
          });
          return {
            timeframeSelectorError: error,
            calendarError: error,
            dateTimeRangePickerError: error
          };
        },
        [formFieldMessagesDispatch]
      )
    );
    const mergedRef = useMergeRefs([
      timeframeSelectorRef,
      forwardedRef
    ]);
    return /* @__PURE__ */ jsx(TimeframeSelectorStoreProvider, { store, children: /* @__PURE__ */ jsx(
      TimeframeSelectorRoot,
      {
        ...remainingProps,
        id,
        ref: mergedRef,
        "aria-disabled": ariaDisabled,
        disabled,
        readOnly
      }
    ) });
  }
);
_TimeframeSelector.displayName = "TimeframeSelector";
const TimeframeSelector = Object.assign(_TimeframeSelector, {
  Trigger,
  DisplayValue,
  CustomTrigger,
  Presets,
  PresetItem
});
export {
  TimeframeSelector
};
