// @dynatrace/strato-components/esm/filters/timeframe-selector/TimeframeSelectorRoot.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import {
  Children,
  forwardRef,
  useCallback,
  useEffect,
  useImperativeHandle,
  useMemo,
  useRef
} from "react";
import { useIntl } from "react-intl";
import { TimeframeSelectorOverlay } from "./components/overlay/TimeframeSelectorOverlay.js";
import { PresetItem } from "./components/presets/PresetItem.js";
import { Presets } from "./components/presets/Presets.js";
import { StepperButton } from "./components/stepper/StepperButton.js";
import { TimeframeSelectorTooltip } from "./components/TimeframeSelectorTooltip.js";
import {
  isTimeframeSelectorCustomTriggerElement
} from "./components/trigger/CustomTrigger.js";
import {
  isTimeframeSelectorTriggerElement,
  Trigger
} from "./components/trigger/Trigger.js";
import { TIMEFRAMESELECTOR_ERROR_MESSAGES } from "./constants/timeframeSelectorMessages.js";
import { TimeframeSelectorContext } from "./contexts/TimeframeSelectorContext.js";
import * as styles from "./TimeframeSelector.sty.js";
import { getTimeframeDisplayValueProps } from "./utils/getTimeframeDisplayValueProps.js";
import { getTimeframePresets } from "./utils/presets-helpers.js";
import {
  TIMEFRAME_PRESETS_QUERY_KEY,
  timeframeQueryClient
} from "./utils/TimeframePresetsClient.js";
import { VisuallyHidden } from "../../core/components/visually-hidden/VisuallyHidden.js";
import { useFormControlState } from "../../core/hooks/useFormControlState.js";
import { useId } from "../../core/hooks/useId.js";
import { useMergeRefs } from "../../core/hooks/useMergeRefs.js";
import { useTimeframeSelectorStore } from "../../core/time/timeframeStore/useTimeframeStore.js";
import { validate } from "../../core/time/validation.js";
import { isSlot } from "../../core/utils/is-slot.js";
import { devLog } from "../../core/utils/logging.js";
import { InternalBaseItem as BaseItem } from "../../forms/forms-core/components/form-field-messages/FormFieldMessageBaseItem.js";
import { useFormFieldMessagesContext } from "../../forms/forms-core/hooks/useFormFieldMessagesContext.js";
import { useFormFieldValidationAriaProps } from "../../forms/forms-core/hooks/useFormFieldValidation.js";
import { useMinMaxLogger } from "../../forms/forms-core/hooks/useMinMaxLogger.js";
import { FormFieldMessagesAction } from "../../forms/forms-core/state/formfield-messages-state-reducer.js";
import {
  parseValue,
  removePrecision
} from "../../forms/forms-core/time/timeParser.js";
import { Flex } from "../../layouts/flex/Flex.js";
import { InputGroup } from "../../layouts/input-group/InputGroup.js";
import { useOverlayWithTrigger } from "../../overlays/hooks/useOverlayWithTrigger.js";
const COMPONENT_NAME = "TimeframeSelector";
function getItemSlots(children, precision) {
  const presetItemsDisplayValuesMap = /* @__PURE__ */ new Map();
  const slots = {
    trigger: null,
    customTrigger: null,
    presets: null,
    presetItemsDisplayValuesMap
  };
  for (const child of Children.toArray(children)) {
    if (isSlot(child, Presets)) {
      slots.presets = child;
      for (const childItem of Children.toArray(child.props.children)) {
        if (isSlot(childItem, PresetItem)) {
          slots.presetItemsDisplayValuesMap.set(
            `${removePrecision(
              childItem.props.value.from,
              precision
            )} ${removePrecision(childItem.props.value.to, precision)}`,
            childItem.props.children
          );
        }
      }
    } else if (isTimeframeSelectorCustomTriggerElement(child)) {
      slots.customTrigger = child;
    } else if (isTimeframeSelectorTriggerElement(child)) {
      slots.trigger = child;
    } else {
      devLog("The used child is not supported by the TimeframeSelector.");
    }
  }
  return slots;
}
const TimeframeSelectorRoot = forwardRef((props, forwardedRef) => {
  const intl = useIntl();
  const {
    disabled,
    value: propValue,
    defaultValue,
    /* eslint-disable @typescript-eslint/no-deprecated */
    clearable = false,
    readOnly = false,
    controlState,
    children,
    onChange,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    name,
    id,
    stepper = true,
    "aria-disabled": ariaDisabled,
    "aria-label": ariaLabelProp,
    "aria-labelledby": ariaLabelledByProp,
    "aria-describedby": ariaDescribedByProp,
    "aria-details": ariaDetailsProp,
    ...remainingProps
  } = props;
  const value = useTimeframeSelectorStore(
    (state) => state.timeframeSelectorTimeframe
  );
  const precision = useTimeframeSelectorStore((state) => state.precision);
  const min = useTimeframeSelectorStore((state) => state.min);
  const max = useTimeframeSelectorStore((state) => state.max);
  const required = useTimeframeSelectorStore((state) => state.required);
  const formatString = useTimeframeSelectorStore((state) => state.formatString);
  const timeframeSelectorRef = useTimeframeSelectorStore(
    (state) => state.timeframeSelectorRef
  );
  const setDateTimeRangePickerTimeframe = useTimeframeSelectorStore(
    (state) => state.setDateTimeRangePickerTimeframe
  );
  const placeholder = intl.formatMessage({
    defaultMessage: "Select timeframe",
    id: "W+rTy8UuFKDho3mc",
    description: "Placeholder shown for the timeframe selector."
  });
  useMinMaxLogger(min, max, "TimeframeSelector");
  const [formControlStateError, hint] = useFormControlState(controlState);
  const { state: formFieldMessagesState, dispatch: formFieldMessagesDispatch } = useFormFieldMessagesContext();
  const onOpenChange = useCallback(
    (isOpen) => {
      if (!isOpen) {
        formFieldMessagesDispatch?.({
          type: FormFieldMessagesAction.SETISTOUCHED,
          payload: { isTouched: true }
        });
        setDateTimeRangePickerTimeframe(value);
      }
    },
    [formFieldMessagesDispatch, setDateTimeRangePickerTimeframe, value]
  );
  const {
    targetRef,
    overlayRef,
    triggerRef,
    overlayProps,
    overlayState,
    overlayTriggerProps
  } = useOverlayWithTrigger({
    offset: 4,
    placement: "bottom-start",
    placementFallbacks: ["bottom-end", "top-start", "top-end"],
    heightStrategy: "maxViewport",
    disabled,
    onOpenChange
  });
  useEffect(() => {
    if (timeframeQueryClient.getQueryData(TIMEFRAME_PRESETS_QUERY_KEY)) {
      return;
    }
    timeframeQueryClient.prefetchQuery({
      queryKey: TIMEFRAME_PRESETS_QUERY_KEY,
      queryFn: getTimeframePresets
    }).catch(() => void 0);
  }, []);
  const details = useMemo(() => {
    const { from, to } = parseValue(value);
    return from && to ? { from, to } : null;
  }, [value]);
  const { customTrigger, trigger, presets, presetItemsDisplayValuesMap } = useMemo(() => getItemSlots(children, precision), [children, precision]);
  const presetItemName = presetItemsDisplayValuesMap.get(
    `${details?.from?.value} ${details?.to?.value}`
  );
  const {
    displayValue,
    isInvalid: displayValueError,
    hint: displayValueHint
  } = getTimeframeDisplayValueProps(
    value,
    intl,
    placeholder,
    precision,
    presetItemName,
    required,
    min?.value,
    max?.value
  );
  const triggerErrorId = useId();
  const inputId = useId();
  const imperativeTriggerRef = useRef(null);
  const containerRef = useRef(null);
  const fromInputRef = useRef(null);
  const toInputRef = useRef(null);
  const ariaLabelingProps = {
    "aria-label": ariaLabelProp,
    "aria-labelledby": ariaLabelledByProp,
    "aria-describedby": ariaDescribedByProp,
    "aria-details": ariaDetailsProp
  };
  const ariaFormProps = useFormFieldValidationAriaProps(ariaLabelingProps);
  useImperativeHandle(
    forwardedRef,
    () => ({
      element: containerRef.current,
      triggerRef: imperativeTriggerRef.current,
      inputRef: { from: fromInputRef.current, to: toInputRef.current },
      open: overlayState.open,
      close: overlayState.close,
      focus() {
        imperativeTriggerRef.current?.focus?.();
      },
      validate() {
        formFieldMessagesDispatch?.({
          type: FormFieldMessagesAction.SETISTOUCHED,
          payload: { isTouched: true }
        });
        validate(value, min, max, {
          formMessageDispatch: formFieldMessagesDispatch,
          precision,
          showError: false,
          checkOrder: true,
          ref: timeframeSelectorRef,
          expectedFormat: formatString,
          valueMode: "timeframeSelector",
          required,
          errorMessages: TIMEFRAMESELECTOR_ERROR_MESSAGES
        });
      }
    }),
    [formFieldMessagesDispatch, overlayState.close, overlayState.open]
  );
  useEffect(() => {
    if (overlayState.isOpen) {
      return;
    }
    if (!displayValueError && triggerErrorId) {
      formFieldMessagesDispatch?.({
        type: FormFieldMessagesAction.REMOVEMESSAGE,
        payload: {
          id: triggerErrorId,
          messageType: "formcontrol",
          controlId: inputId
        }
      });
    }
  }, [
    displayValueError,
    displayValueHint,
    overlayState.isOpen,
    triggerErrorId,
    formFieldMessagesDispatch,
    inputId
  ]);
  const triggerButtonRef = useMergeRefs([
    targetRef,
    triggerRef,
    imperativeTriggerRef
  ]);
  const handleOnBlur = useCallback(() => {
    if (!overlayState.isOpen) {
      fromInputRef.current?.reportValidity();
      formFieldMessagesDispatch?.({
        type: FormFieldMessagesAction.SETISTOUCHED,
        payload: { isTouched: true }
      });
    }
  }, [overlayState.isOpen, formFieldMessagesDispatch]);
  const triggerProps = useMemo(
    () => ({
      disabled,
      onBlur: handleOnBlur,
      ref: triggerButtonRef,
      id,
      "aria-disabled": ariaDisabled,
      ...ariaFormProps,
      ...readOnly ? {} : overlayTriggerProps,
      "data-dt-component": COMPONENT_NAME
    }),
    [
      disabled,
      handleOnBlur,
      triggerButtonRef,
      id,
      ariaDisabled,
      ariaFormProps,
      readOnly,
      overlayTriggerProps
    ]
  );
  const timeframeSelectorContextValue = useMemo(
    () => ({
      details,
      displayValue,
      triggerProps,
      hint: displayValueHint,
      readOnly,
      showError: formControlStateError || formFieldMessagesState.showError,
      disabled
    }),
    [
      disabled,
      details,
      readOnly,
      displayValue,
      triggerProps,
      displayValueHint,
      formControlStateError,
      formFieldMessagesState.showError
    ]
  );
  return /* @__PURE__ */ jsx(TimeframeSelectorContext.Provider, { value: timeframeSelectorContextValue, children: /* @__PURE__ */ jsxs(
    Flex,
    {
      gap: 4,
      flexDirection: "column",
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ref: containerRef,
      className: clsx(consumerClassName, styles.root),
      style: consumerStyle,
      ...remainingProps,
      "data-dt-component": "TimeframeSelector",
      children: [
        /* @__PURE__ */ jsxs(VisuallyHidden, { children: [
          /* @__PURE__ */ jsx(
            "input",
            {
              id: inputId,
              type: "text",
              ref: fromInputRef,
              name: name && `${name}-from`,
              "aria-hidden": true,
              readOnly,
              required,
              disabled,
              value: details?.from.value ?? "",
              onInvalid: (event) => {
                formFieldMessagesDispatch?.({
                  type: FormFieldMessagesAction.SETISTOUCHED,
                  payload: { isTouched: true }
                });
                event.preventDefault();
              },
              onChange: () => {
              },
              tabIndex: -1,
              ...ariaFormProps
            }
          ),
          /* @__PURE__ */ jsx(
            "input",
            {
              type: "text",
              ref: toInputRef,
              name: name && `${name}-to`,
              "aria-hidden": true,
              readOnly,
              required,
              disabled,
              value: details?.to.value ?? "",
              onInvalid: (event) => event.preventDefault(),
              onChange: () => {
              },
              tabIndex: -1,
              ...ariaFormProps
            }
          )
        ] }),
        /* @__PURE__ */ jsxs(InputGroup, { style: { gridAutoColumns: "1fr max-content max-content" }, children: [
          /* @__PURE__ */ jsx(
            TimeframeSelectorTooltip,
            {
              displayValue,
              value,
              disabled,
              children: customTrigger ?? trigger ?? /* @__PURE__ */ jsx(Trigger, {})
            }
          ),
          stepper && /* @__PURE__ */ jsxs(Fragment, { children: [
            /* @__PURE__ */ jsx(
              StepperButton,
              {
                disabled: disabled || readOnly,
                direction: "backward",
                "aria-disabled": ariaDisabled,
                onBlur: handleOnBlur
              }
            ),
            /* @__PURE__ */ jsx(
              StepperButton,
              {
                disabled: disabled || readOnly,
                direction: "forward",
                "aria-disabled": ariaDisabled,
                onBlur: handleOnBlur
              }
            )
          ] })
        ] }),
        controlState && hint && /* @__PURE__ */ jsx(
          BaseItem,
          {
            variant: formControlStateError ? "error" : "hint",
            disabled: disabled || !!ariaDisabled,
            children: hint
          }
        ),
        overlayState.isOpen && /* @__PURE__ */ jsx(
          TimeframeSelectorOverlay,
          {
            ref: overlayRef,
            overlayProps,
            overlayClose: overlayState.close,
            clearable,
            "data-dtrum-mask": dataDtrumMask,
            "data-dtrum-allow": dataDtrumAllow,
            children: presets
          }
        )
      ]
    }
  ) });
});
export {
  TimeframeSelectorRoot
};
