// @dynatrace/strato-components/esm/charts/histogram/constants/defaults.js
import {
  DEFAULT_TRUNCATION_MODE,
  DEFAULT_VALUE_REPRESENTATION,
  defaultRatio,
  SCALE_FALLBACK_COLOR
} from "../../core/constants.js";
const DISPLAY_NAME = "HistogramChart";
const DEFAULT_CHART_HEIGHT = 300;
const DEFAULT_LOADING_STATE = false;
const defaultLegend = {
  hidden: false,
  position: "auto",
  ratio: defaultRatio
};
const defaultColorPalette = "categorical";
const DEFAULT_SCALE_FALLBACK_COLOR = SCALE_FALLBACK_COLOR;
const defaultTooltipVariant = "single";
const defaultToolbarPlacement = "top-right";
const defaultToolbar = {
  placement: defaultToolbarPlacement,
  enabled: true
};
const defaultInteractions = {
  zoom: { enabled: true },
  pan: { enabled: true }
};
const defaultHistogramChartConfig = {
  valueRepresentation: DEFAULT_VALUE_REPRESENTATION,
  legend: defaultLegend,
  colorPalette: defaultColorPalette,
  tooltip: {
    variant: defaultTooltipVariant
  },
  truncationMode: DEFAULT_TRUNCATION_MODE,
  toolbar: defaultToolbar,
  interactions: defaultInteractions
};
const defaultYAxis = {};
const defaultXAxis = {
  scale: "linear"
};
const defaultThreshold = {
  leftThreshold: [],
  rightThreshold: []
};
const HISTOGRAM_CHART_ARIA_LABEL = "Histogram Chart";
export {
  DEFAULT_CHART_HEIGHT,
  DEFAULT_LOADING_STATE,
  DEFAULT_SCALE_FALLBACK_COLOR,
  DISPLAY_NAME,
  HISTOGRAM_CHART_ARIA_LABEL,
  defaultColorPalette,
  defaultHistogramChartConfig,
  defaultLegend,
  defaultThreshold,
  defaultToolbar,
  defaultToolbarPlacement,
  defaultTooltipVariant,
  defaultXAxis,
  defaultYAxis
};
