// @dynatrace/strato-design-tokens/esm/animations-js/index.js
const Open = {
  Duration: {
    /** Value: 0.17. */
    Fast: 0.17,
    /** Value: 0.25. */
    Base: 0.25,
    /** Value: 0.4. */
    Slow: 0.4
  },
  /** Value: [ 0, 0,0.2,1 ]. */
  Easing: [0, 0, 0.2, 1]
};
const Close = {
  Duration: {
    /** Value: 0. */
    Fast: 0,
    /** Value: 0.17. */
    Base: 0.17,
    /** Value: 0.25. */
    Slow: 0.25
  },
  /** Value: [ 0.5, 0,1,1 ]. */
  Easing: [0.5, 0, 1, 1]
};
const SlideIn = {
  Duration: {
    /** Value: 0.2. */
    Fast: 0.2,
    /** Value: 0.3. */
    Base: 0.3,
    /** Value: 0.4. */
    Slow: 0.4
  },
  /** Value: [ 0, 0,0.2,1 ]. */
  Easing: [0, 0, 0.2, 1]
};
const SlideOut = {
  Duration: {
    /** Value: 0.15. */
    Fast: 0.15,
    /** Value: 0.25. */
    Base: 0.25,
    /** Value: 0.35. */
    Slow: 0.35
  },
  /** Value: [ 0.8, 0,0.5,1 ]. */
  Easing: [0.8, 0, 0.5, 1]
};
const FadeIn = {
  Duration: {
    /** Value: 0.15. */
    Fast: 0.15,
    /** Value: 0.25. */
    Base: 0.25,
    /** Value: 0.35. */
    Slow: 0.35
  },
  /** Value: [ 0.5, 0,1,1 ]. */
  Easing: [0.5, 0, 1, 1]
};
const FadeOut = {
  Duration: {
    /** Value: 0.15. */
    Fast: 0.15,
    /** Value: 0.25. */
    Base: 0.25,
    /** Value: 0.35. */
    Slow: 0.35
  },
  /** Value: [ 0, 0,0.2,1 ]. */
  Easing: [0, 0, 0.2, 1]
};
const Hover = {
  Duration: {
    /** Value: 0.15. */
    Base: 0.15
  },
  /** Value: [ 0, 0,0.2,1 ]. */
  Easing: [0, 0, 0.2, 1]
};
const StaggeredMotion = {
  Duration: {
    /** Value: 0.25. */
    Fast: 0.25,
    /** Value: 0.35. */
    Base: 0.35,
    /** Value: 0.4. */
    Slow: 0.4
  },
  Offset: {
    /** Value: 0.02. */
    Fast: 0.02,
    /** Value: 0.03. */
    Base: 0.03,
    /** Value: 0.05. */
    Slow: 0.05
  },
  /** Value: [ 0, 0,0.2,1 ]. */
  Easing: [0, 0, 0.2, 1]
};
const ExpandCollapse = {
  Duration: {
    /** Value: 0.25. */
    Fast: 0.25,
    /** Value: 0.4. */
    Base: 0.4,
    /** Value: 0.6. */
    Slow: 0.6
  },
  /** Value: [ 0.4, 0,0.2,1 ]. */
  Easing: [0.4, 0, 0.2, 1]
};
const Move = {
  Duration: {
    /** Value: 0.15. */
    Fast: 0.15,
    /** Value: 0.4. */
    Base: 0.4,
    /** Value: 0.6. */
    Slow: 0.6
  },
  /** Value: [ 0.4, 0,0.2,1 ]. */
  Easing: [0.4, 0, 0.2, 1]
};
const Rescale = {
  Duration: {
    /** Value: 0.17. */
    Fast: 0.17,
    /** Value: 0.25. */
    Base: 0.25,
    /** Value: 0.4. */
    Slow: 0.4
  },
  /** Value: [ 0.4, 0,0.2,1 ]. */
  Easing: [0.4, 0, 0.2, 1]
};
var animations_js_default = {
  Open,
  Close,
  SlideIn,
  SlideOut,
  FadeIn,
  FadeOut,
  Hover,
  StaggeredMotion,
  ExpandCollapse,
  Move,
  Rescale
};
export {
  animations_js_default as default
};
