// @dynatrace/strato-components/esm/layouts/page-layout/sidebar-layout/CollapsedSidebarRail.js
import { jsx, jsxs } from "react/jsx-runtime";
import { ChevronSmallRightDoubleIcon } from "@dynatrace/strato-icons";
import * as styles from "./SidebarLayout.sty.js";
import { Button } from "../../../buttons/button/Button.js";
import { Tooltip } from "../../../overlays/tooltip/Tooltip.js";
function SidebarHint({
  panelId,
  expandSidebarLabel,
  expandButtonRef,
  onCollapsedChange,
  isCollapsed = true
}) {
  const expand = () => {
    onCollapsedChange(false);
  };
  return /* @__PURE__ */ jsxs("div", { className: styles.collapsedSidebar, id: panelId, children: [
    /* @__PURE__ */ jsx(
      "button",
      {
        className: styles.hint,
        type: "button",
        onClick: expand,
        tabIndex: -1,
        "aria-hidden": "true",
        "aria-label": expandSidebarLabel
      }
    ),
    /* @__PURE__ */ jsx("div", { className: styles.expandButtonContainer, children: /* @__PURE__ */ jsx(Tooltip, { text: expandSidebarLabel, children: /* @__PURE__ */ jsx(
      Button,
      {
        className: styles.expandButton,
        "aria-expanded": !isCollapsed,
        ref: expandButtonRef,
        "aria-controls": panelId,
        "aria-label": expandSidebarLabel,
        size: "condensed",
        "data-dt-name": "ExpandCollapse",
        "data-dt-properties": "state:expand",
        onClick: expand,
        children: /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(ChevronSmallRightDoubleIcon, {}) })
      }
    ) }) })
  ] });
}
export {
  SidebarHint
};
