// @dynatrace/strato-components/esm/layouts/page-layout/sidebar-layout/SidebarLayout.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { clsx } from "clsx";
import {
  forwardRef,
  memo,
  useRef
} from "react";
import { useIntl } from "react-intl";
import { ChevronSmallLeftDoubleIcon, XmarkIcon } from "@dynatrace/strato-icons";
import { SidebarHint } from "./CollapsedSidebarRail.js";
import * as styles from "./SidebarLayout.sty.js";
import { useButtonFocusHandover } from "./useFocusHandover.js";
import { useLayoutState } from "./useLayoutState.js";
import { Button } from "../../../buttons/button/Button.js";
import { useId } from "../../../core/hooks/useId.js";
import { useMergeRefs } from "../../../core/hooks/useMergeRefs.js";
import { Drawer } from "../../../overlays/drawer/Drawer.js";
import { Tooltip } from "../../../overlays/tooltip/Tooltip.js";
import { SplitLayout } from "../../split-layout/SplitLayout.js";
import { useTeleport } from "../shared/useTeleport.js";
function SidebarLayoutInner({
  sidebar,
  content,
  containerRef,
  panelId,
  persistedPanelWidthRef,
  contentTeleport,
  panelTeleport
}) {
  const layout = useLayoutState({ sidebar, containerRef });
  const intl = useIntl();
  const focusHandover = useButtonFocusHandover();
  const expandSidebarLabel = intl.formatMessage({
    defaultMessage: "Expand sidebar",
    id: "NF0tLfJTokD/xw5B",
    description: "Label for the button that expands the sidebar."
  });
  const collapseSidebarLabel = intl.formatMessage({
    defaultMessage: "Collapse sidebar",
    id: "LJ0fyD2++VZgZ5g0",
    description: "Label for the button that collapses the sidebar."
  });
  const sidebarLabel = intl.formatMessage({
    defaultMessage: "Sidebar",
    id: "BYUJpW8TBonAxnYd",
    description: "Accessible label for the sidebar panel."
  });
  const setCollapsedWithFocusHandover = (value) => {
    layout.setCollapsed(value);
    focusHandover.setTarget(value ? "expand" : "collapse");
  };
  switch (layout.mode) {
    case "collapsed": {
      return /* @__PURE__ */ jsxs("div", { className: styles.collapsedSidebarWrapper, children: [
        /* @__PURE__ */ jsx(
          SidebarHint,
          {
            panelId,
            expandSidebarLabel,
            expandButtonRef: focusHandover.expandButtonRef,
            onCollapsedChange: setCollapsedWithFocusHandover
          }
        ),
        /* @__PURE__ */ jsx(
          "div",
          {
            ref: content.ref,
            ...content.domProps,
            className: clsx(styles.content, content.domProps.className),
            children: /* @__PURE__ */ jsx(contentTeleport.Target, {})
          }
        )
      ] });
    }
    case "split": {
      const splitLayoutKey = persistedPanelWidthRef.current === void 0 ? `split-${String(sidebar.config.defaultWidth)}` : "split-resized";
      return /* @__PURE__ */ jsxs(SplitLayout, { children: [
        /* @__PURE__ */ jsx(
          SplitLayout.Panel,
          {
            ...sidebar.domProps,
            ref: sidebar.ref,
            minWidth: sidebar.config.minWidth,
            maxWidth: sidebar.config.maxWidth,
            defaultWidth: persistedPanelWidthRef.current ?? sidebar.config.defaultWidth,
            resizable: sidebar.config.resizable,
            position: "start",
            onResize: (value) => {
              persistedPanelWidthRef.current = value;
              sidebar.config.onResize?.(value);
            },
            className: clsx(styles.panel, sidebar.domProps.className),
            "aria-label": sidebarLabel,
            id: panelId,
            children: /* @__PURE__ */ jsx(panelTeleport.Target, {})
          }
        ),
        /* @__PURE__ */ jsx(SplitLayout.ResizeHandle, { className: styles.resizeHandle, children: /* @__PURE__ */ jsx("div", { className: styles.collapseButtonContainer, children: /* @__PURE__ */ jsx(Tooltip, { text: collapseSidebarLabel, children: /* @__PURE__ */ jsx(
          Button,
          {
            className: styles.collapseButton,
            "aria-expanded": "true",
            ref: focusHandover.collapseButtonRef,
            "aria-controls": panelId,
            "aria-label": collapseSidebarLabel,
            size: "condensed",
            "data-dt-name": "ExpandCollapse",
            "data-dt-properties": "state:collapse",
            onClick: () => {
              setCollapsedWithFocusHandover(true);
            },
            children: /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(ChevronSmallLeftDoubleIcon, {}) })
          }
        ) }) }) }),
        /* @__PURE__ */ jsx(
          SplitLayout.Content,
          {
            ...content.domProps,
            ref: content.ref,
            minWidth: content.config.minWidth,
            className: clsx(styles.content, content.domProps.className),
            children: /* @__PURE__ */ jsx(contentTeleport.Target, {})
          }
        )
      ] }, splitLayoutKey);
    }
    case "drawer": {
      const drawerContentArea = /* @__PURE__ */ jsx(
        "div",
        {
          ref: content.ref,
          ...content.domProps,
          className: clsx(styles.content, content.domProps.className),
          children: /* @__PURE__ */ jsx(contentTeleport.Target, {})
        }
      );
      return /* @__PURE__ */ jsxs(Fragment, { children: [
        /* @__PURE__ */ jsxs("div", { className: styles.collapsedSidebarWrapper, children: [
          /* @__PURE__ */ jsx(
            SidebarHint,
            {
              panelId,
              expandSidebarLabel,
              expandButtonRef: focusHandover.expandButtonRef,
              isCollapsed: layout.collapsed,
              onCollapsedChange: setCollapsedWithFocusHandover
            }
          ),
          drawerContentArea
        ] }),
        /* @__PURE__ */ jsx(
          Drawer,
          {
            ...sidebar.domProps,
            ref: sidebar.ref,
            className: clsx(styles.drawer, sidebar.domProps.className),
            "aria-label": sidebarLabel,
            isDismissed: layout.collapsed,
            onDismiss: (state, _reason) => {
              setCollapsedWithFocusHandover(state);
            },
            placement: "left",
            width: "wide",
            children: /* @__PURE__ */ jsxs("div", { className: styles.drawerContentContainer, children: [
              /* @__PURE__ */ jsx("div", { className: styles.drawerHeader, children: /* @__PURE__ */ jsx(
                Button,
                {
                  style: { marginLeft: "auto" },
                  ref: focusHandover.collapseButtonRef,
                  "aria-label": collapseSidebarLabel,
                  "data-dt-name": "CloseX",
                  onClick: () => {
                    setCollapsedWithFocusHandover(true);
                  },
                  children: /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(XmarkIcon, {}) })
                }
              ) }),
              /* @__PURE__ */ jsx("div", { className: styles.drawerContent, children: /* @__PURE__ */ jsx(panelTeleport.Target, {}) })
            ] })
          }
        )
      ] });
    }
  }
}
const SidebarLayout = memo(
  forwardRef((props, ref) => {
    const { className: rootClassName, sidebar, content, ...rootProps } = props;
    const panelId = useId();
    const containerRef = useRef(null);
    const rootRef = useMergeRefs([ref, containerRef]);
    const persistedPanelWidth = useRef();
    const contentTeleport = useTeleport();
    const panelTeleport = useTeleport();
    return /* @__PURE__ */ jsxs(Fragment, { children: [
      /* @__PURE__ */ jsx(contentTeleport.Source, { children: content.children }),
      sidebar && /* @__PURE__ */ jsx(panelTeleport.Source, { children: sidebar.children }),
      /* @__PURE__ */ jsx(
        "div",
        {
          ...rootProps,
          ref: rootRef,
          className: clsx(styles.layoutRoot, rootClassName),
          children: sidebar ? /* @__PURE__ */ jsx(
            SidebarLayoutInner,
            {
              sidebar,
              content,
              containerRef,
              panelId,
              persistedPanelWidthRef: persistedPanelWidth,
              contentTeleport,
              panelTeleport
            }
          ) : /* @__PURE__ */ jsx(
            "div",
            {
              ref: content.ref,
              ...content.domProps,
              className: clsx(styles.content, content.domProps.className),
              children: /* @__PURE__ */ jsx(contentTeleport.Target, {})
            }
          )
        }
      )
    ] });
  })
);
SidebarLayout.displayName = "SidebarLayout (forwardRef)";
export {
  SidebarLayout
};
