// @dynatrace/strato-components/esm/layouts/page-layout/sidebar-layout/SidebarLayout.sty.js
import "../../../core/components/app-root/AppRoot.css";
import "../../../core/styles/focusRingStatic.css";
import "./SidebarLayout.css";
var collapseButton = "_1mr4xg7f-3-6-0 _1mr4xg7e-3-6-0";
var collapseButtonContainer = "_1mr4xg7c-3-6-0";
var collapsedSidebar = "_1mr4xg75-3-6-0";
var collapsedSidebarWrapper = "_1mr4xg71-3-6-0";
var content = "_1mr4xg73-3-6-0";
var drawer = "_1mr4xg77-3-6-0";
var drawerContent = "_1mr4xg79-3-6-0";
var drawerContentContainer = "_1mr4xg78-3-6-0";
var drawerHeader = "_1mr4xg7a-3-6-0";
var expandButton = "_1mr4xg7g-3-6-0 _1mr4xg7e-3-6-0";
var expandButtonContainer = "_1mr4xg7d-3-6-0";
var hint = "_1mr4xg76-3-6-0 _z0fq9b4-3-6-0 _z0fq9bd-3-6-0 _z0fq9b6-3-6-0 _z0fq9bc-3-6-0";
var layoutRoot = "_1mr4xg70-3-6-0";
var panel = "_1mr4xg72-3-6-0";
var resizeHandle = "_1mr4xg74-3-6-0";
export {
  collapseButton,
  collapseButtonContainer,
  collapsedSidebar,
  collapsedSidebarWrapper,
  content,
  drawer,
  drawerContent,
  drawerContentContainer,
  drawerHeader,
  expandButton,
  expandButtonContainer,
  hint,
  layoutRoot,
  panel,
  resizeHandle
};
