// @dynatrace/strato-components/esm/content/release-phase-indicator/ReleasePhaseIndicator.js
import { jsx } from "react/jsx-runtime";
import { forwardRef } from "react";
import { useIntl } from "react-intl";
import { releasePhaseMessages } from "./messages.js";
import { useAriaLabelingProps } from "../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useSafeForwardProps } from "../../core/hooks/useSafeForwardProps.js";
import { Chip } from "../chip/Chip.js";
const COMPONENT_NAME = "ReleasePhaseIndicator";
const phaseColorMap = {
  preview: "primary",
  new: "primary",
  deprecated: "warning"
};
const phaseLabelMap = {
  preview: releasePhaseMessages.preview,
  new: releasePhaseMessages.new,
  deprecated: releasePhaseMessages.deprecated
};
const ReleasePhaseIndicator = forwardRef((props, forwardedRef) => {
  const {
    phase,
    id,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    ...remainingProps
  } = props;
  const intl = useIntl();
  const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
  const ariaLabelingProps = useAriaLabelingProps(restProps);
  const forwardProps = useSafeForwardProps(restProps, ariaLabelingProps);
  return /* @__PURE__ */ jsx(
    Chip,
    {
      id,
      ref: forwardedRef,
      size: "condensed",
      variant: "emphasized",
      color: phaseColorMap[phase],
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...ariaLabelingProps,
      ...behavioralTrackingProps,
      ...forwardProps,
      "data-dt-component": COMPONENT_NAME,
      children: intl.formatMessage(phaseLabelMap[phase])
    }
  );
});
ReleasePhaseIndicator.displayName = "ReleasePhaseIndicator";
export {
  ReleasePhaseIndicator
};
