// @dynatrace/strato-components/forms/spin-buttons/number-spin-button/NumberSpinButton.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var NumberSpinButton_exports = {};
__export(NumberSpinButton_exports, {
  NumberSpinButton: () => NumberSpinButton,
  isNumberSpinButton: () => isNumberSpinButton
});
module.exports = __toCommonJS(NumberSpinButton_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_is = require("react-is");
var import_useBeforeInputHandler = require("./hooks/useBeforeInputHandler.js");
var import_useValidationAndFormatting = require("./hooks/useValidationAndFormatting.js");
var import_useValueNavigation = require("./hooks/useValueNavigation.js");
var import_useAriaLabelingProps = require("../../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../../core/hooks/useBehavioralTrackingProps.js");
var import_merge_props = require("../../../core/utils/merge-props.js");
var import_useSelection = require("../hooks/useSelection.js");
var import_SpinButton_css = require("../SpinButton.sty.js");
var import_constants = require("./constants.js");
var import_useSpinButtonInteraction = require("../hooks/useSpinButtonInteraction.js");
const NumberSpinButton = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    disabled = false,
    readOnly = false,
    value: valueProp,
    placeholder,
    className: consumerClassName,
    style: consumerStyle,
    displayFormat,
    required,
    "data-testid": dataTestId,
    onChange,
    defaultValue = import_constants.NUMBER_SPIN_BUTTON_DEFAULTS.VALUE,
    minValue,
    maxValue,
    hasError,
    onKeyDownHandler: propKeyDownHandler,
    isCyclic = false,
    "data-pattern": dataPattern,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, remainingPropsWithoutTracking] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const [text, setText] = (0, import_react.useState)("");
  const [value, setValue] = (0, import_react.useState)(defaultValue);
  const [error, setError] = (0, import_react.useState)(false);
  const ariaProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(props);
  const rootRef = (0, import_react.useRef)(null);
  const { getSelection, setSelection, selectAll, deselectAll } = (0, import_useSelection.useSelection)(rootRef);
  const { validate, format, parse } = (0, import_useValidationAndFormatting.useValidationAndFormatting)(
    displayFormat,
    required,
    minValue ?? import_constants.NUMBER_SPIN_BUTTON_DEFAULTS.MIN_VALUE,
    maxValue ?? import_constants.NUMBER_SPIN_BUTTON_DEFAULTS.MAX_VALUE
  );
  const applyText = (0, import_react.useCallback)(
    (input, fireOnChange = false) => {
      const newValue = parse(input);
      setText(input);
      setValue(newValue);
      if (rootRef?.current) {
        rootRef.current.textContent = format(input);
      }
      fireOnChange && onChange?.(newValue);
    },
    [format, onChange, parse, setValue]
  );
  const clearValue = (0, import_react.useCallback)(
    (element, fireOnChange = false) => {
      setValue(null);
      setText("");
      if (element) {
        element.textContent = "";
      }
      fireOnChange && onChange?.(null);
    },
    [onChange, setValue]
  );
  const beforeInputHandler = (0, import_useBeforeInputHandler.useBeforeInputHandler)(
    rootRef,
    props,
    applyText,
    text
  );
  const { onKeyDownHandler } = (0, import_useValueNavigation.useValueNavigation)(
    rootRef,
    {
      value: valueProp ?? value,
      disabled,
      readOnly,
      onChange,
      displayFormat,
      required,
      minValue,
      maxValue,
      isCyclic,
      ...remainingPropsWithoutTracking
    },
    applyText
  );
  (0, import_react.useImperativeHandle)(
    forwardedRef,
    () => ({
      type: "number",
      element: rootRef,
      setSelection,
      getSelection,
      selectAll,
      deselectAll,
      applyText: (input, fireOnChange) => applyText(input, fireOnChange),
      clearValue: (fireOnChange) => clearValue(rootRef.current, fireOnChange)
    }),
    [setSelection, getSelection, selectAll, deselectAll, clearValue, applyText]
  );
  (0, import_react.useEffect)(() => {
    const isValid = validate(value);
    setError(!isValid);
  }, [setError, validate, value]);
  (0, import_react.useEffect)(() => {
    applyText(
      defaultValue === null || defaultValue === void 0 || Number.isNaN(defaultValue) ? "" : defaultValue.toString()
    );
  }, [format]);
  (0, import_react.useEffect)(() => {
    const element = rootRef.current;
    element?.addEventListener("beforeinput", beforeInputHandler);
    return () => element?.removeEventListener("beforeinput", beforeInputHandler);
  }, [beforeInputHandler]);
  (0, import_react.useEffect)(() => {
    if (valueProp === void 0) {
      return;
    }
    const sel = getSelection();
    if (valueProp !== value) {
      applyText(
        valueProp !== null && Number.isInteger(valueProp) ? valueProp.toString() : ""
      );
    }
    document.activeElement === rootRef.current && sel && setSelection(sel);
  }, [valueProp, applyText, format, getSelection, setSelection]);
  const { onClick, onFocus, onBlur } = (0, import_useSpinButtonInteraction.useSpinButtonInteraction)({
    disabled,
    selectAll,
    deselectAll,
    rootElement: rootRef.current
  });
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "span",
    {
      "data-testid": dataTestId,
      ref: rootRef,
      tabIndex: disabled ? void 0 : 0,
      contentEditable: (
        // contentEditable when focused and non-editable when blurred, allowing text selection.
        !disabled && !readOnly ? true : void 0
      ),
      spellCheck: false,
      className: (0, import_clsx.default)(
        consumerClassName,
        (0, import_SpinButton_css.spinButtonCSS)({ disabled, error: hasError ?? error })
      ),
      style: {
        "--data-placeholder": `"${placeholder ?? ""}"`,
        ...consumerStyle
      },
      "aria-disabled": disabled,
      ...(0, import_merge_props.mergeProps)(ariaProps, behavioralTrackingProps),
      "aria-label": ariaProps["aria-label"] ?? placeholder,
      role: "spinbutton",
      onClick,
      onFocus,
      onBlur,
      onKeyDown: propKeyDownHandler ?? onKeyDownHandler,
      "data-pattern": dataPattern
    }
  );
});
function isNumberSpinButton(node) {
  return node !== void 0 && node !== null && (0, import_react_is.isElement)(node) && node.type === NumberSpinButton;
}
NumberSpinButton.displayName = "_NumberSpinButton";
