// @dynatrace/strato-components/esm/charts/core/components/threshold-bar/ThresholdBarTemplate.js
import { jsx, jsxs } from "react/jsx-runtime";
import { useIntl } from "react-intl";
import { useThresholdBarTooltip } from "./hooks/useThresholdBarTooltip.js";
import { CircleMarker } from "./markers/CircleMarker.js";
import { MARKER_OPACITY, MARKER_SHADOW_OPACITY } from "./markers/constants.js";
import { PillMarker } from "./markers/PillMarker.js";
import { wrapper } from "./styles.sty.js";
import { THRESHOLD_BAR_THICKNESS } from "./types/constants.js";
import { getThresholdIntValue } from "./utils/get-threshold-int-value.js";
import { resolveInfinitePixel } from "./utils/resolve-infinite-pixel.js";
import { sortMarkers } from "./utils/sort-circle-markers.js";
import { OverlayTooltip } from "../overlay-tooltip/OverlayTooltip.js";
const ThresholdBarTemplate = (props) => {
  const {
    borderArea,
    length,
    markers,
    layout = "vertical",
    isFlipped = false,
    onHover
  } = props;
  const sortedMarkers = markers.toSorted(sortMarkers);
  const intl = useIntl();
  const {
    showTooltip,
    tooltipPosition,
    tooltipItems,
    handleMouseMove,
    handleMouseLeave
  } = useThresholdBarTooltip(sortedMarkers, onHover);
  const flipType = isFlipped ? "flip" : "no-flip";
  const layoutType = `${layout}-${flipType}`;
  return /* @__PURE__ */ jsxs(
    "div",
    {
      className: wrapper({
        borderArea,
        layoutType
      }),
      style: { height: length },
      children: [
        /* @__PURE__ */ jsx("svg", { width: THRESHOLD_BAR_THICKNESS, height: length, children: sortedMarkers.map((marker, index) => {
          const { y0, y1, color, isShadow = false } = marker;
          const isMarkerARange = y0 !== y1;
          const resolvedY0 = resolveInfinitePixel(y0, length);
          const resolvedY1 = resolveInfinitePixel(y1, length);
          if (isMarkerARange) {
            return [
              /* @__PURE__ */ jsx(
                PillMarker,
                {
                  x: 4,
                  y0: resolvedY0,
                  y1: resolvedY1,
                  color,
                  opacity: isShadow ? MARKER_SHADOW_OPACITY : MARKER_OPACITY,
                  onMouseMove: (event) => handleMouseMove(event, index),
                  onMouseLeave: handleMouseLeave
                },
                `threshold-bar-pill-marker-${marker.id}`
              )
            ];
          } else {
            return [
              /* @__PURE__ */ jsx(
                CircleMarker,
                {
                  x: 8,
                  y: resolvedY0,
                  color,
                  onMouseMove: (event) => handleMouseMove(event, index),
                  onMouseLeave: handleMouseLeave
                },
                `threshold-bar-circle-marker-${marker.id}`
              )
            ];
          }
        }) }),
        showTooltip && tooltipPosition && /* @__PURE__ */ jsx(
          OverlayTooltip,
          {
            isInteractive: true,
            absoluteX: tooltipPosition.x,
            absoluteY: tooltipPosition.y,
            children: tooltipItems.map(({ id, name, color, values }) => {
              return /* @__PURE__ */ jsx(
                OverlayTooltip.SeriesItem,
                {
                  color: color.Default,
                  name,
                  value: `${getThresholdIntValue(intl, values)}`
                },
                id
              );
            })
          }
        )
      ]
    }
  );
};
ThresholdBarTemplate["displayName"] = "ThresholdBarTemplate";
export {
  ThresholdBarTemplate
};
