// @dynatrace/strato-components/content/message-container/MessageContainer.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MessageContainer_exports = {};
__export(MessageContainer_exports, {
  MessageContainer: () => MessageContainer
});
module.exports = __toCommonJS(MessageContainer_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_Actions = require("./Actions.js");
var import_MessageContainerContext = require("./contexts/MessageContainerContext.js");
var import_Description = require("./Description.js");
var import_Prefix = require("./Prefix.js");
var import_Title = require("./Title.js");
var import_Button = require("../../buttons/button/Button.js");
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useSafeForwardProps = require("../../core/hooks/useSafeForwardProps.js");
var import_useSlots = require("../../core/hooks/useSlots.js");
var import_logging = require("../../core/utils/logging.js");
var import_Container = require("../../layouts/container/Container.js");
var import_Flex = require("../../layouts/flex/Flex.js");
const COMPONENT_NAME = "MessageContainer";
const _MessageContainer = (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const {
      children,
      variant = "neutral",
      onDismiss,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-allow": dataDtrumAllow,
      "data-dtrum-mask": dataDtrumMask,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
    const ariaLabelingProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(restProps);
    const forwardProps = (0, import_useSafeForwardProps.useSafeForwardProps)(restProps, ariaLabelingProps);
    const intl = (0, import_react_intl.useIntl)();
    const [slots] = (0, import_useSlots.useSlots)(children, {
      prefix: import_Prefix.Prefix,
      title: import_Title.Title,
      description: import_Description.Description,
      actions: import_Actions.Actions
    });
    if (!slots.description) {
      (0, import_logging.devLog)(
        "The MessageContainer component requires a Description component."
      );
    }
    const wrapActions = slots.title && slots.description;
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_MessageContainerContext.MessageContainerContext.Provider, { value: { variant }, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Container.Container,
      {
        color: variant,
        padding: 6,
        paddingLeft: 16,
        ref: forwardedRef,
        className: consumerClassName,
        style: consumerStyle,
        "data-testid": dataTestId,
        "data-dtrum-allow": dataDtrumAllow,
        "data-dtrum-mask": dataDtrumMask,
        ...ariaLabelingProps,
        ...behavioralTrackingProps,
        ...forwardProps,
        "data-dt-component": COMPONENT_NAME,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { justifyContent: "space-between", alignItems: "flex-start", children: [
          slots.prefix && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Flex.Flex, { flex: "0 0 20px", paddingY: 6, justifyContent: "center", children: slots.prefix }),
          /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
            import_Flex.Flex,
            {
              flexGrow: 1,
              columnGap: 8,
              rowGap: 6,
              flexDirection: wrapActions ? "column" : "row",
              flexWrap: "wrap",
              minWidth: 0,
              children: [
                /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
                  import_Flex.Flex,
                  {
                    flexDirection: "column",
                    gap: 2,
                    flexGrow: 1,
                    justifyContent: "center",
                    paddingY: 6,
                    paddingRight: 8,
                    maxWidth: "100%",
                    minWidth: 0,
                    style: { boxSizing: "border-box" },
                    children: [
                      slots.title,
                      slots.description
                    ]
                  }
                ),
                slots.actions && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  import_Flex.Flex,
                  {
                    ...wrapActions && {
                      style: { paddingRight: 8, paddingBottom: 10 }
                    },
                    children: slots.actions
                  }
                )
              ]
            }
          ),
          onDismiss !== void 0 && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Flex.Flex, { flexShrink: 0, alignItems: "flex-start", children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_Button.Button,
            {
              onClick: onDismiss,
              color: variant,
              "aria-label": intl.formatMessage({
                defaultMessage: "Close message",
                id: "2+ypcBK9NuQ5zg1k",
                description: "Aria label used for the close button of the information"
              }),
              children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.XmarkIcon, {}) })
            }
          ) })
        ] })
      }
    ) });
  }
);
_MessageContainer.displayName = COMPONENT_NAME;
const MessageContainer = Object.assign(_MessageContainer, {
  Prefix: import_Prefix.Prefix,
  Title: import_Title.Title,
  Description: import_Description.Description,
  Actions: import_Actions.Actions
});
