// @dynatrace/strato-components/filters/timeframe-selector/components/stepper/StepperButtonTooltip.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var StepperButtonTooltip_exports = {};
__export(StepperButtonTooltip_exports, {
  StepperButtonTooltip: () => StepperButtonTooltip
});
module.exports = __toCommonJS(StepperButtonTooltip_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_Flex = require("../../../../layouts/flex/Flex.js");
var import_Tooltip = require("../../../../overlays/tooltip/Tooltip.js");
var import_Text = require("../../../../typography/text/Text.js");
var import_useTooltipMessages = require("../../hooks/useTooltipMessages.js");
const StepperButtonTooltip = (props) => {
  const {
    value,
    disabled,
    children,
    formattedStepperValue
  } = props;
  const { tooltipFormattedTimeValueStart, tooltipFormattedTimeValueEnd } = (0, import_useTooltipMessages.useTooltipMessages)({
    from: value?.from.value ?? "",
    to: value?.to.value ?? ""
  });
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Tooltip.Tooltip,
    {
      text: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { flexDirection: "column", gap: 0, children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { textStyle: "small-emphasized", children: formattedStepperValue }),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { textStyle: "small", children: tooltipFormattedTimeValueStart }),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { textStyle: "small", children: tooltipFormattedTimeValueEnd })
      ] }),
      placement: "top",
      fallbackPlacements: ["top-end", "bottom", "bottom-end", "right", "left"],
      disabled,
      children
    }
  );
};
StepperButtonTooltip.displayName = "TimeframeStepperTooltip";
