// @dynatrace/strato-components/filters/timeframe-selector/components/stepper/StepperButton.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var StepperButton_exports = {};
__export(StepperButton_exports, {
  StepperButton: () => StepperButton
});
module.exports = __toCommonJS(StepperButton_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_StepperButtonTooltip = require("./StepperButtonTooltip.js");
var import_Button = require("../../../../buttons/button/Button.js");
var import_helpers = require("../../../../core/time/helpers.js");
var import_useTimeframeStore = require("../../../../core/time/timeframeStore/useTimeframeStore.js");
var import_useTimeframeSelectorContext = require("../../hooks/useTimeframeSelectorContext.js");
var import_stepper_helpers = require("../../utils/stepper-helpers.js");
var import_timeframe_selection = require("../../utils/timeframe-selection.js");
function isStepperDisabled(direction, value, min, max) {
  if (!value) {
    return false;
  }
  if (direction === "backward") {
    const nextStepIsValid = (0, import_timeframe_selection.isSmallerThan)(value.from.absoluteDate, min) || (0, import_timeframe_selection.dateIsMinMax)(value.from, min);
    return nextStepIsValid;
  }
  if (direction === "forward") {
    const nextStepIsValid = (0, import_timeframe_selection.isBiggerThan)(value.to.absoluteDate, max) || (0, import_timeframe_selection.dateIsMinMax)(value.to, void 0, max);
    return nextStepIsValid;
  }
  return false;
}
const StepperButton = (props) => {
  const {
    disabled,
    direction,
    "aria-disabled": ariaDisabledProp,
    onBlur,
    readOnly
  } = props;
  const { showError } = (0, import_useTimeframeSelectorContext.useTimeframeSelectorContext)();
  const precision = (0, import_useTimeframeStore.useTimeframeSelectorStore)((state) => state.precision);
  const min = (0, import_useTimeframeStore.useTimeframeSelectorStore)((state) => state.min)?.value;
  const max = (0, import_useTimeframeStore.useTimeframeSelectorStore)((state) => state.max)?.value;
  const isFullDay = (0, import_useTimeframeStore.useTimeframeSelectorStore)((state) => state.isFullDay);
  const setTimeframeSelectorTimeframe = (0, import_useTimeframeStore.useTimeframeSelectorStore)(
    (state) => state.setStepperValue
  );
  const value = (0, import_useTimeframeStore.useTimeframeSelectorStore)(
    (state) => (0, import_helpers.isFullTimeframe)(state.timeframeSelectorTimeframe) ? state.timeframeSelectorTimeframe : null
  );
  const referenceTimeframe = (0, import_useTimeframeStore.useTimeframeSelectorStore)(
    (state) => state.referenceTimeframe
  );
  const setReferenceTimeframe = (0, import_useTimeframeStore.useTimeframeSelectorStore)(
    (state) => state.setReferenceTimeframe
  );
  const [shiftValue, setShiftValue] = (0, import_react.useState)(value);
  const [isDisabled, setIsDisabled] = (0, import_react.useState)(
    isStepperDisabled(direction, value, min, max)
  );
  (0, import_react.useEffect)(() => {
    setIsDisabled(isStepperDisabled(direction, value, min, max));
    const newShiftValue = (0, import_stepper_helpers.getTimeShift)(value, {
      direction,
      min,
      max,
      selectedOverCalendar: isFullDay,
      precision
    });
    if (newShiftValue) {
      setShiftValue(newShiftValue);
    }
  }, [direction, value, min, max, isFullDay, isDisabled, precision]);
  const handleShift = (0, import_react.useCallback)(() => {
    const newShiftValue = (0, import_stepper_helpers.getTimeShift)(referenceTimeframe ?? value, {
      direction,
      min,
      max,
      selectedOverCalendar: isFullDay,
      precision
    });
    if (newShiftValue && !readOnly) {
      const isAtMinOrMax = direction === "backward" ? (0, import_timeframe_selection.dateIsMinMax)(newShiftValue?.from, min, max) : (0, import_timeframe_selection.dateIsMinMax)(newShiftValue?.to, min, max);
      if (referenceTimeframe && (0, import_helpers.isFullTimeframe)(referenceTimeframe)) {
        setTimeframeSelectorTimeframe(referenceTimeframe);
        setReferenceTimeframe(null);
        return;
      }
      if (isAtMinOrMax) {
        setReferenceTimeframe(value);
      }
      setTimeframeSelectorTimeframe(newShiftValue);
    }
  }, [
    value,
    direction,
    min,
    max,
    isFullDay,
    precision,
    readOnly,
    referenceTimeframe,
    setReferenceTimeframe,
    setTimeframeSelectorTimeframe
  ]);
  const intl = (0, import_react_intl.useIntl)();
  const dataDtName = direction === "backward" ? "StepBackward" : "StepForward";
  if (!value || !shiftValue) {
    return null;
  }
  const formattedStepperValue = (0, import_stepper_helpers.formatStepperValues)(
    shiftValue.from,
    shiftValue.to,
    intl,
    direction,
    isFullDay
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_StepperButtonTooltip.StepperButtonTooltip,
    {
      disabled: disabled || isDisabled,
      value: shiftValue,
      formattedStepperValue,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_Button.Button,
        {
          style: { minWidth: "fit-content" },
          variant: "emphasized",
          color: showError ? "critical" : "neutral",
          disabled: disabled || isDisabled,
          onClick: handleShift,
          "aria-disabled": ariaDisabledProp,
          "aria-label": formattedStepperValue,
          onBlur,
          "data-dt-name": dataDtName,
          children: direction === "backward" ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ChevronLeftSmallIcon, {}) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ChevronRightSmallIcon, {})
        }
      )
    }
  );
};
StepperButton.displayName = "TimeframeStepperButton";
