// @dynatrace/strato-components/charts/core/components/axes/utils/get-tick-label-props.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var get_tick_label_props_exports = {};
__export(get_tick_label_props_exports, {
  ELLIPSIS_BUFFER: () => ELLIPSIS_BUFFER,
  bottomPanTickLabelDateProps: () => bottomPanTickLabelDateProps,
  bottomTickLabelDateProps: () => bottomTickLabelDateProps,
  getBottomTickLabelPropsWithDomainAlignment: () => getBottomTickLabelPropsWithDomainAlignment,
  getTickLabelPropsForCategoryAxis: () => getTickLabelPropsForCategoryAxis,
  getTickLabelPropsForGanttAxis: () => getTickLabelPropsForGanttAxis,
  getTickLabelPropsForValueAxis: () => getTickLabelPropsForValueAxis,
  getTopTickLabelPropsWithDomainAlignment: () => getTopTickLabelPropsWithDomainAlignment,
  getXAxisTickLabelProps: () => getXAxisTickLabelProps
});
module.exports = __toCommonJS(get_tick_label_props_exports);
var import_get_tick_value_as_number = require("../../../utils/get-tick-value-as-number.js");
var import_constants = require("../constants.js");
const getBottomTickLabelPropsWithDomainAlignment = (domainMin, domainMax) => {
  return (value) => {
    const numericValue = (0, import_get_tick_value_as_number.getTickValueAsNumber)(value);
    let textAnchor = import_constants.TICK_ALIGMENT.Middle;
    if (numericValue === domainMin) {
      textAnchor = import_constants.TICK_ALIGMENT.Start;
    } else if (numericValue === domainMax) {
      textAnchor = import_constants.TICK_ALIGMENT.End;
    }
    return {
      y: import_constants.X_AXES_TICK_LINE_OFFSET,
      textAnchor,
      verticalAnchor: import_constants.TICK_ALIGMENT.Start
    };
  };
};
const VISX_BOTTOM_OFFSET = 2;
const ELLIPSIS_BUFFER = 2;
const leftTickLabelProps = (_value, index, values) => ({
  x: -import_constants.Y_AXES_TICK_LINE_OFFSET,
  textAnchor: import_constants.TICK_ALIGMENT.End,
  verticalAnchor: index === 0 ? import_constants.TICK_ALIGMENT.End : index === values.length - 1 ? import_constants.TICK_ALIGMENT.Start : import_constants.TICK_ALIGMENT.Middle
});
const rightTickLabelProps = (_value, index, values) => ({
  x: +import_constants.Y_AXES_TICK_LINE_OFFSET,
  textAnchor: import_constants.TICK_ALIGMENT.Start,
  verticalAnchor: index === 0 ? import_constants.TICK_ALIGMENT.End : index === values.length - 1 ? import_constants.TICK_ALIGMENT.Start : import_constants.TICK_ALIGMENT.Middle
});
const leftTickCategoricalLabelProps = (_value) => ({
  x: -import_constants.Y_AXES_TICK_LINE_OFFSET,
  textAnchor: import_constants.TICK_ALIGMENT.End,
  verticalAnchor: import_constants.TICK_ALIGMENT.Middle
});
const topTickLabelProps = (_value, index, values) => ({
  y: -import_constants.X_AXES_TICK_LINE_OFFSET,
  textAnchor: index === 0 ? import_constants.TICK_ALIGMENT.Start : index === values.length - 1 ? import_constants.TICK_ALIGMENT.End : import_constants.TICK_ALIGMENT.Middle,
  verticalAnchor: import_constants.TICK_ALIGMENT.End
});
const topRotatedTickLabelProps = (_value, index, values) => ({
  y: import_constants.X_AXES_TICK_LINE_OFFSET,
  textAnchor: index === 0 ? import_constants.TICK_ALIGMENT.Start : index === values.length - 1 ? import_constants.TICK_ALIGMENT.End : import_constants.TICK_ALIGMENT.Middle,
  verticalAnchor: import_constants.TICK_ALIGMENT.Middle,
  angle: -90
});
const topTickLabelPropsForCategoryAxis = (_value) => ({
  y: import_constants.X_AXES_TICK_LINE_OFFSET,
  textAnchor: import_constants.TICK_ALIGMENT.Middle,
  verticalAnchor: import_constants.TICK_ALIGMENT.End
});
const topRotatedTickLabelPropsForCategoryAxis = (_value) => ({
  y: import_constants.X_AXES_TICK_LINE_OFFSET,
  textAnchor: import_constants.TICK_ALIGMENT.End,
  verticalAnchor: import_constants.TICK_ALIGMENT.Middle,
  angle: -90
});
const bottomTickLabelProps = (_value, index, values) => ({
  y: import_constants.X_AXES_TICK_LINE_OFFSET,
  textAnchor: index === 0 ? import_constants.TICK_ALIGMENT.Start : index === values.length - 1 ? import_constants.TICK_ALIGMENT.End : import_constants.TICK_ALIGMENT.Middle,
  verticalAnchor: import_constants.TICK_ALIGMENT.Start
});
const bottomRotatedTickLabelPropsForCategoryAxis = (_value) => ({
  y: import_constants.X_AXES_TICK_LINE_OFFSET,
  textAnchor: import_constants.TICK_ALIGMENT.End,
  verticalAnchor: import_constants.TICK_ALIGMENT.Middle,
  angle: -90
});
const bottomRotatedTickLabelProps = (_value, index, values) => ({
  y: import_constants.X_AXES_TICK_LINE_OFFSET,
  textAnchor: index === 0 ? import_constants.TICK_ALIGMENT.Start : index === values.length - 1 ? import_constants.TICK_ALIGMENT.End : import_constants.TICK_ALIGMENT.Middle,
  verticalAnchor: import_constants.TICK_ALIGMENT.Middle,
  angle: -90
});
const bottomTickLabelPropsForCategoryAxis = (_value) => ({
  y: import_constants.X_AXES_TICK_LINE_OFFSET,
  textAnchor: import_constants.TICK_ALIGMENT.Middle,
  verticalAnchor: import_constants.TICK_ALIGMENT.Start
});
const getTopTickLabelPropsWithDomainAlignment = (domainMin, domainMax) => {
  return (value) => {
    const numericValue = (0, import_get_tick_value_as_number.getTickValueAsNumber)(value);
    let textAnchor = import_constants.TICK_ALIGMENT.Middle;
    if (numericValue === domainMin) {
      textAnchor = import_constants.TICK_ALIGMENT.Start;
    } else if (numericValue === domainMax) {
      textAnchor = import_constants.TICK_ALIGMENT.End;
    }
    return {
      y: -import_constants.X_AXES_TICK_LINE_OFFSET,
      textAnchor,
      verticalAnchor: import_constants.TICK_ALIGMENT.End
    };
  };
};
const getTickLabelPropsForCategoryAxis = (orientation, tickLabelLayout = "horizontal") => {
  if (orientation === "left") {
    return leftTickCategoricalLabelProps;
  }
  if (orientation === "top") {
    return tickLabelLayout === "vertical" ? topRotatedTickLabelPropsForCategoryAxis : topTickLabelPropsForCategoryAxis;
  }
  return tickLabelLayout === "vertical" ? bottomRotatedTickLabelPropsForCategoryAxis : bottomTickLabelPropsForCategoryAxis;
};
const bottomTickLabelDateProps = (_value) => ({
  y: import_constants.X_AXES_TICK_LINE_OFFSET + VISX_BOTTOM_OFFSET,
  textAnchor: import_constants.TICK_ALIGMENT.Middle,
  verticalAnchor: import_constants.TICK_ALIGMENT.Start
});
const bottomPanTickLabelDateProps = (_value, index, values) => {
  const result = {
    textAnchor: import_constants.TICK_ALIGMENT.Middle,
    verticalAnchor: import_constants.TICK_ALIGMENT.Middle
  };
  const isFirstItem = index === 0;
  const isLastItem = index === values.length - 1;
  if (isFirstItem) {
    result.textAnchor = import_constants.TICK_ALIGMENT.Start;
  }
  if (isLastItem) {
    result.textAnchor = import_constants.TICK_ALIGMENT.End;
  }
  return result;
};
const getTickLabelPropsForValueAxis = (orientation, tickLabelLayout = "horizontal") => {
  if (orientation === "left") {
    return leftTickLabelProps;
  }
  if (orientation === "right") {
    return rightTickLabelProps;
  }
  if (orientation === "top") {
    return tickLabelLayout === "vertical" ? topRotatedTickLabelProps : topTickLabelProps;
  }
  return tickLabelLayout === "vertical" ? bottomRotatedTickLabelProps : bottomTickLabelProps;
};
function getXAxisTickLabelProps(position, isCategoricalAxis, isTickRotated, scale, tickLabelLayout) {
  if (isCategoricalAxis) {
    return getTickLabelPropsForCategoryAxis(position, tickLabelLayout);
  }
  if (scale && !isTickRotated) {
    const domain = scale.domain();
    const getDomainAligned = position === "bottom" ? getBottomTickLabelPropsWithDomainAlignment : getTopTickLabelPropsWithDomainAlignment;
    return getDomainAligned(+domain[0], +domain[1]);
  }
  return getTickLabelPropsForValueAxis(position, tickLabelLayout);
}
const getTickLabelPropsForGanttAxis = (min, max) => {
  return (value, index, values) => {
    let textAnchor = import_constants.TICK_ALIGMENT.Middle;
    let dx = 0;
    const valueAsNumber = (0, import_get_tick_value_as_number.getTickValueAsNumber)(value);
    if (values.length === 1 && valueAsNumber && min && max) {
      textAnchor = valueAsNumber - min < valueAsNumber - max ? import_constants.TICK_ALIGMENT.Start : import_constants.TICK_ALIGMENT.End;
    } else if (index === 0) {
      textAnchor = import_constants.TICK_ALIGMENT.Start;
      dx = valueAsNumber === min ? 4 : 0;
    } else if (index === values.length - 1) {
      textAnchor = import_constants.TICK_ALIGMENT.End;
      dx = valueAsNumber === max ? -4 : 0;
    }
    return {
      dx,
      dy: -2,
      textAnchor,
      verticalAnchor: import_constants.TICK_ALIGMENT.Middle
    };
  };
};
