// @dynatrace/strato-components/tables/DataTable/features/Sorting/compare-functions.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var compare_functions_exports = {};
__export(compare_functions_exports, {
  alphanumericCompare: () => alphanumericCompare,
  datelikeCompare: () => datelikeCompare,
  numberlikeCompare: () => numberlikeCompare,
  stringlikeCompare: () => stringlikeCompare,
  stringlikeCompareCaseSensitive: () => stringlikeCompareCaseSensitive
});
module.exports = __toCommonJS(compare_functions_exports);
var import_string_utils = require("../../../../core/utils/string-utils.js");
function convertToSafeNumber(value) {
  const convertedValue = typeof value === "number" ? value : (0, import_string_utils.convertToNumberIfPureNumber)(value);
  const isValueBigInt = value !== Infinity && value !== -Infinity && typeof convertedValue === "number" && (convertedValue > Number.MAX_SAFE_INTEGER || convertedValue < Number.MIN_SAFE_INTEGER);
  return isValueBigInt ? BigInt(value) : convertedValue;
}
function numberlikeCompare(valueA, valueB) {
  let aValue = valueA;
  let bValue = valueB;
  if (typeof valueA === "string" || typeof valueA === "number") {
    aValue = convertToSafeNumber(valueA);
  }
  if (typeof valueB === "string" || typeof valueB === "number") {
    bValue = convertToSafeNumber(valueB);
  }
  if ((typeof aValue === "number" || typeof aValue === "bigint") && (typeof bValue === "number" || typeof bValue === "bigint")) {
    return compareBasic(aValue, bValue);
  }
  return alphanumericCompare(
    convertToString(aValue).toLowerCase(),
    convertToString(bValue).toLowerCase()
  );
}
function alphanumericCompare(valueA, valueB) {
  const reSplitAlphaNumeric = /([0-9]+)/gm;
  const aValue = valueA.split(reSplitAlphaNumeric).filter(Boolean);
  const bValue = valueB.split(reSplitAlphaNumeric).filter(Boolean);
  while (aValue.length && bValue.length) {
    const aValueString = aValue.shift();
    const bValueString = bValue.shift();
    const aValueInt = parseInt(aValueString, 10);
    const bValueInt = parseInt(bValueString, 10);
    const combo = [aValueInt, bValueInt].sort();
    if (isNaN(combo[0])) {
      if (aValueString > bValueString) {
        return 1;
      }
      if (bValueString > aValueString) {
        return -1;
      }
      continue;
    }
    if (isNaN(combo[1])) {
      return isNaN(aValueInt) ? -1 : 1;
    }
    if (aValueInt > bValueInt) {
      return 1;
    }
    if (bValueInt > aValueInt) {
      return -1;
    }
  }
  return aValue.length - bValue.length;
}
function stringlikeCompare(valueA, valueB) {
  return compareBasic(
    convertToString(valueA).toLowerCase(),
    convertToString(valueB).toLowerCase()
  );
}
function stringlikeCompareCaseSensitive(valueA, valueB) {
  return compareBasic(valueA, valueB);
}
function convertToNanoseconds(value) {
  if (Number.isNaN(value)) {
    return NaN;
  }
  if (typeof value === "number") {
    return BigInt(value) * 1000000n;
  }
  if (value instanceof Date) {
    const valTime = value.getTime();
    if (Number.isNaN(valTime)) {
      return NaN;
    }
    return BigInt(valTime) * 1000000n;
  }
  if (typeof value === "string") {
    return getPreciseTime(value);
  }
  return NaN;
}
function datelikeCompare(valueA, valueB) {
  const aValue = convertToNanoseconds(valueA);
  const bValue = convertToNanoseconds(valueB);
  if (Number.isNaN(aValue)) {
    return 1;
  }
  if (Number.isNaN(bValue)) {
    return -1;
  }
  return numberlikeCompare(aValue, bValue);
}
function compareBasic(valueA, valueB) {
  return valueA === valueB ? 0 : valueA > valueB ? 1 : -1;
}
function convertToString(value) {
  if (value instanceof Date) {
    return value.toISOString();
  }
  if (typeof value === "number") {
    return String(value);
  }
  if (typeof value === "string") {
    return value;
  }
  return "";
}
function getPreciseTime(time) {
  const milliseconds = new Date(time).getTime();
  if (isNaN(milliseconds)) {
    return NaN;
  }
  const regex = /\.(\d{1,9})/;
  const fractionalMatch = regex.exec(time)?.[1] ?? "";
  const microAndNanoseconds = fractionalMatch.padEnd(9, "0").substring(3, 9);
  const msToNs = BigInt(milliseconds) * 1000000n;
  const remainingNs = BigInt(microAndNanoseconds);
  return msToNs + remainingNs;
}
