// @dynatrace/strato-components/tables/DataTable/features/Sorting/Sorting.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Sorting_exports = {};
__export(Sorting_exports, {
  Sorting: () => Sorting,
  useSorting: () => useSorting
});
module.exports = __toCommonJS(Sorting_exports);
var import_react_table = require("@tanstack/react-table");
var import_compare_functions = require("./compare-functions.js");
var import_useControllableState = require("../../../../core/hooks/useControllableState.js");
var import_useDataTableConfig = require("../../../hooks/useDataTableConfig.js");
var import_datatable_typeguards = require("../../hooks/useTable/datatable-typeguards.js");
var import_utils = require("../../hooks/useTable/utils.js");
function isColumnDefWithSortAccessor(columnDef) {
  return columnDef !== void 0 && columnDef !== null && typeof columnDef === "object" && "sortAccessor" in columnDef && columnDef.sortAccessor !== void 0;
}
function useSorting(props, options) {
  const configSortable = (0, import_useDataTableConfig.useDataTableConfigElement)("sortable");
  const [sorting, setSorting] = (0, import_useControllableState.useControllableState)({
    value: props.sortBy,
    defaultValue: props.defaultSortBy,
    onChange: props.onSortByChange
  });
  const { sortable: sortableProp } = props;
  let sortable = false;
  if (sortableProp !== void 0) {
    if (typeof sortableProp === "boolean") {
      sortable = sortableProp;
      options.manualSorting = false;
    } else {
      sortable = true;
      if (sortableProp.manualSort === true) {
        options.manualSorting = true;
      } else {
        options.manualSorting = void 0;
      }
    }
  } else if (configSortable !== void 0) {
    if (typeof configSortable === "boolean") {
      sortable = configSortable;
    } else {
      sortable = true;
      if (configSortable.manualSort === true) {
        options.manualSorting = true;
      } else {
        options.manualSorting = void 0;
      }
    }
  }
  options.enableSorting = sortable;
  options.state = {
    ...options.state,
    sorting
  };
  options.onSortingChange = setSorting;
}
const sortingFunctions = {
  text: import_compare_functions.stringlikeCompare,
  textCaseSensitive: import_compare_functions.stringlikeCompareCaseSensitive,
  number: import_compare_functions.numberlikeCompare,
  datetime: import_compare_functions.datelikeCompare
};
const Sorting = {
  createColumn(column, table) {
    column.getCustomSortingFn = () => {
      if (!column) {
        throw new Error();
      }
      if ((0, import_react_table.isFunction)(column.columnDef.customSortingFn)) {
        return column.columnDef.customSortingFn;
      }
      if (typeof column.columnDef.sortType === "string" && sortingFunctions[column.columnDef.sortType]) {
        return sortingFunctions[column.columnDef.sortType];
      }
      return sortingFunctions["text"];
    };
    column.getAutoSortDir = () => {
      return "asc";
    };
    column.getCanSort = () => {
      if ((0, import_datatable_typeguards.isGroupColumnDef)(column.columnDef)) {
        return false;
      }
      return (column.columnDef.enableSorting ?? true) && (table.options.enableSorting ?? false) && !!column.accessorFn;
    };
    const originalToggleSorting = column.toggleSorting;
    column.toggleSorting = (desc, isMulti) => {
      if (!isMulti) {
        const oldSorting = table.getState().sorting;
        if (oldSorting && oldSorting.length > 1) {
          const existingIndex = oldSorting.findIndex((s) => s.id === column.id);
          if (existingIndex === oldSorting.length - 1) {
            const nextSortingOrder = column.getNextSortingOrder();
            const nextDesc = desc ?? nextSortingOrder === "desc";
            table.setSorting([{ id: column.id, desc: nextDesc }]);
            return;
          }
        }
      }
      originalToggleSorting(desc, isMulti);
    };
  },
  createTable(table) {
    (0, import_utils.assignRowMethods)(table, {
      getSortingValue(columnId) {
        if (Object.hasOwn(this._sortingValuesCache, columnId)) {
          return this._sortingValuesCache[columnId];
        }
        const column = table.getColumn(columnId);
        if (column && isColumnDefWithSortAccessor(column.columnDef)) {
          const sortAccessor = column.columnDef.sortAccessor;
          if (sortAccessor && typeof sortAccessor === "string") {
            return sortAccessor.split(".").reduce(
              // @ts-expect-error Strict mode transition - needs to be investigated
              (obj, key) => obj[key],
              this.original
            );
          } else if (typeof sortAccessor === "function") {
            return sortAccessor(this.original);
          }
        }
        if (!column?.accessorFn) {
          return void 0;
        }
        this._sortingValuesCache[columnId] = column.accessorFn(
          this.original,
          this.index
        );
        return this._sortingValuesCache[columnId];
      }
    });
  },
  createRow(row, table) {
    row._sortingValuesCache = {};
  },
  createHeader(header, table) {
    header.getAriaSortAttribute = () => {
      if (table.options.enableSorting && header.column.columnDef.enableSorting !== false && header.column.columns.length === 0) {
        const isSorted = header.column.getIsSorted();
        if (isSorted) {
          return isSorted === "asc" ? "ascending" : "descending";
        }
        return "none";
      }
      return void 0;
    };
    header.getCanSortByHeaderClick = () => {
      if (header.column.getCanColumnActions() === true) {
        return false;
      }
      return header.column.getCanSort();
    };
  }
};
