// @dynatrace/strato-components/tables/DataTable/features/Sorting/sorted-row-model.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var sorted_row_model_exports = {};
__export(sorted_row_model_exports, {
  getSortedRowModel: () => getSortedRowModel
});
module.exports = __toCommonJS(sorted_row_model_exports);
var import_utils = require("../../hooks/useTable/utils.js");
var import_src = require("../../tanstack-table-core-override/src/index.js");
function getSortedRowModel() {
  return (table) => (0, import_src.memo)(
    () => [table.getState().sorting, table.getPreSortedRowModel()],
    (sorting, rowModel) => {
      if (!rowModel.rows.length || !sorting?.length) {
        return rowModel;
      }
      const sortingState = table.getState().sorting;
      const sortedFlatRows = [];
      const availableSorting = sortingState.filter(
        (sort) => table.getColumn(sort.id)?.getCanSort()
      );
      const columnInfoById = {};
      availableSorting.forEach((sortEntry) => {
        const column = table.getColumn(sortEntry.id);
        if (!column) {
          return;
        }
        columnInfoById[sortEntry.id] = {
          sortUndefined: column.columnDef.sortUndefined,
          invertSorting: column.columnDef.invertSorting,
          columnType: column.columnDef.columnType,
          customSortingFn: column.getCustomSortingFn()
        };
      });
      const sortData = (rows) => {
        const sortedData = rows.map((row) => row.clone());
        sortedData.sort((rowA, rowB) => {
          for (let i = 0; i < availableSorting.length; i += 1) {
            const sortEntry = availableSorting[i];
            const columnInfo = columnInfoById[sortEntry.id];
            const isDesc = sortEntry?.desc ?? false;
            let sortInt = 0;
            if (columnInfo.sortUndefined) {
              const aValue = rowA.getSortingValue(sortEntry.id);
              const bValue = rowB.getSortingValue(sortEntry.id);
              const aUndefined = aValue === void 0 || aValue === null || columnInfo.columnType !== void 0 && Number.isNaN(aValue);
              const bUndefined = bValue === void 0 || bValue === null || columnInfo.columnType !== void 0 && Number.isNaN(bValue);
              if (aUndefined || bUndefined) {
                if (aUndefined && bUndefined) {
                  sortInt = 0;
                }
                const sorting2 = columnInfo.invertSorting && isDesc || !columnInfo.invertSorting && !isDesc;
                if (aUndefined) {
                  sortInt = sorting2 ? 1 : -1;
                } else {
                  sortInt = sorting2 ? -1 : 1;
                }
              }
            }
            if (sortInt === 0) {
              sortInt = columnInfo.customSortingFn(
                rowA.getSortingValue(sortEntry.id),
                rowB.getSortingValue(sortEntry.id)
              );
            }
            if (sortInt !== 0) {
              if (isDesc) {
                sortInt *= -1;
              }
              if (columnInfo.invertSorting) {
                sortInt *= -1;
              }
              return sortInt;
            }
          }
          return rowA.index - rowB.index;
        });
        sortedData.forEach((row) => {
          sortedFlatRows.push(row);
          if (row.subRows?.length) {
            row.subRows = sortData(row.subRows);
          }
        });
        return sortedData;
      };
      return {
        rows: sortData(rowModel.rows),
        flatRows: sortedFlatRows,
        rowsById: rowModel.rowsById
      };
    },
    (0, import_utils.getMemoOptions)(
      table.options,
      "debugTable",
      "getSortedRowModel",
      () => table._autoResetPageIndex()
    )
  );
}
