// @dynatrace/strato-components/tables/DataTable/features/Sorting/SortingArrowIndicators.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SortingArrowIndicators_exports = {};
__export(SortingArrowIndicators_exports, {
  SortingArrowIndicators: () => SortingArrowIndicators
});
module.exports = __toCommonJS(SortingArrowIndicators_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var styles = __toESM(require("./Sorting.sty.js"));
var import_Text = require("../../../../typography/text/Text.js");
var import_messages = require("../../messages.js");
const SortingArrowIndicators = (props) => {
  const { isSorted, multiSortPriority, nextSortingOrder, rowDensity } = props;
  const intl = (0, import_react_intl.useIntl)();
  const dataDtName = "ColumnHeaderSortIcon";
  let dataDtProperties = "state:clearSort";
  if (nextSortingOrder === "asc") {
    dataDtProperties = "state:ascending";
  } else if (nextSortingOrder === "desc") {
    dataDtProperties = "state:descending";
  }
  if (isSorted !== false) {
    const sortedDirectionIcon = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_strato_icons.ArrowSmallDownIcon,
      {
        "data-sorticon": true,
        "data-dt-name": dataDtName,
        "data-dt-properties": dataDtProperties,
        style: { display: "inline" },
        className: (0, import_clsx.default)(
          styles.sortingArrow({ arrowUp: isSorted === "asc", rowDensity })
        )
      }
    );
    if (multiSortPriority !== void 0) {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Text.Text, { as: "span", className: styles.multiSortIndicatorBubble, children: [
        sortedDirectionIcon,
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_Text.Text,
          {
            as: "span",
            textStyle: "small-emphasized",
            "aria-hidden": "true",
            className: styles.sortOrderIndicator,
            children: multiSortPriority
          }
        )
      ] });
    }
    return sortedDirectionIcon;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_strato_icons.SorterSmallDoubleIcon,
    {
      "data-sorticon": true,
      "aria-label": intl.formatMessage(import_messages.dataTableMessages.sortableColumnIcon),
      className: styles.sortingArrow({ rowDensity }),
      "data-dt-name": dataDtName,
      "data-dt-properties": dataDtProperties
    }
  );
};
