// @dynatrace/strato-components/charts/gauge/constants.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var constants_exports = {};
__export(constants_exports, {
  ARC_BACKGROUND_COLOR: () => ARC_BACKGROUND_COLOR,
  ARC_DEFAULT_VALUE_COLOR: () => ARC_DEFAULT_VALUE_COLOR,
  ARC_END_ANGLE: () => ARC_END_ANGLE,
  ARC_INNER_RADIUS_PERCENTAGE: () => ARC_INNER_RADIUS_PERCENTAGE,
  ARC_OUTER_RADIUS_PERCENTAGE: () => ARC_OUTER_RADIUS_PERCENTAGE,
  ARC_START_ANGLE: () => ARC_START_ANGLE,
  DEFAULT_CHART_HEIGHT: () => DEFAULT_CHART_HEIGHT,
  DEFAULT_CHART_WIDTH: () => DEFAULT_CHART_WIDTH,
  DEFAULT_MAX_RANGE_VALUE: () => DEFAULT_MAX_RANGE_VALUE,
  DEFAULT_MAX_VALUE_FONT_SIZE: () => DEFAULT_MAX_VALUE_FONT_SIZE,
  DEFAULT_MIN_RANGE_VALUE: () => DEFAULT_MIN_RANGE_VALUE,
  DISPLAY_NAME: () => DISPLAY_NAME,
  FONT_HEIGHT_RATIO: () => FONT_HEIGHT_RATIO,
  GAUGE_CHART_ARIA_LABEL: () => GAUGE_CHART_ARIA_LABEL,
  GAUGE_DEFAULT_CONFIG: () => GAUGE_DEFAULT_CONFIG,
  GAUGE_VERTICAL_DEFAULT_OFFSET: () => GAUGE_VERTICAL_DEFAULT_OFFSET,
  GAUGE_VERTICAL_SMALL_OFFSET: () => GAUGE_VERTICAL_SMALL_OFFSET,
  MIN_MAX_LABEL_RESERVED_HEIGHT_IN_PX: () => MIN_MAX_LABEL_RESERVED_HEIGHT_IN_PX,
  MIN_MAX_PADDING: () => MIN_MAX_PADDING,
  MIN_SIZE_IN_PX: () => MIN_SIZE_IN_PX,
  THRESHOLD_COLOR_DEFAULT: () => THRESHOLD_COLOR_DEFAULT,
  THRESHOLD_LINE_DEFAULT_COLOR: () => THRESHOLD_LINE_DEFAULT_COLOR,
  THRESHOLD_LINE_WIDTH: () => THRESHOLD_LINE_WIDTH,
  zIndexGaugeAreas: () => zIndexGaugeAreas
});
module.exports = __toCommonJS(constants_exports);
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
const DEFAULT_MIN_RANGE_VALUE = 0;
const DEFAULT_MAX_RANGE_VALUE = 100;
const DEFAULT_CHART_WIDTH = "100%";
const DEFAULT_CHART_HEIGHT = 300;
const MIN_SIZE_IN_PX = 100;
const MIN_MAX_PADDING = 20;
const MIN_MAX_LABEL_RESERVED_HEIGHT_IN_PX = 16;
const ARC_OUTER_RADIUS_PERCENTAGE = 0.9;
const ARC_INNER_RADIUS_PERCENTAGE = 0.8;
const GAUGE_VERTICAL_SMALL_OFFSET = 0.05;
const GAUGE_VERTICAL_DEFAULT_OFFSET = 0.1;
const ARC_BACKGROUND_COLOR = import_colors.default.Theme.Neutral["40"];
const ARC_DEFAULT_VALUE_COLOR = import_colors.default.Charts.Categorical.Color11.Default;
const THRESHOLD_COLOR_DEFAULT = import_colors.default.Charts.Categorical.Color11.Default;
const ARC_START_ANGLE = -0.625 * Math.PI;
const ARC_END_ANGLE = 0.625 * Math.PI;
const THRESHOLD_LINE_WIDTH = 2;
const THRESHOLD_LINE_DEFAULT_COLOR = import_colors.default.Icon.Neutral.Default;
const GAUGE_DEFAULT_CONFIG = {
  color: ARC_DEFAULT_VALUE_COLOR,
  min: DEFAULT_MIN_RANGE_VALUE,
  max: DEFAULT_MAX_RANGE_VALUE,
  formatter: void 0,
  loading: false,
  tooltip: {}
};
const DISPLAY_NAME = "GaugeChart";
const GAUGE_CHART_ARIA_LABEL = "Gauge Chart";
const FONT_HEIGHT_RATIO = 0.7;
const DEFAULT_MAX_VALUE_FONT_SIZE = 240;
const zIndexGaugeAreas = {
  toolbar: 2,
  tooltip: 1,
  overlay: 0
};
