// @dynatrace/strato-components/charts/gauge/GaugeChart.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var GaugeChart_exports = {};
__export(GaugeChart_exports, {
  GaugeChart: () => GaugeChart
});
module.exports = __toCommonJS(GaugeChart_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_useObserveContainerSize = require("../core/hooks/useObserveContainerSize.js");
var import_chart_types = require("../core/types/chart-types.js");
var import_coerceSizeValue = require("../core/utils/coerceSizeValue.js");
var import_GaugeChartLayout = require("./components/GaugeChartLayout/GaugeChartLayout.js");
var import_constants = require("./constants.js");
var import_useGaugeChartConfig = require("./hooks/useGaugeChartConfig.js");
var import_GaugeChartProviders = require("./providers/GaugeChartProviders.js");
var import_Label = require("./slots/Label.js");
var import_ThresholdIndicator = require("./slots/ThresholdIndicator.js");
var import_Toolbar = require("./slots/Toolbar.js");
var import_Tooltip = require("./slots/Tooltip.js");
var import_Intent = require("../../core/slots/Intent/Intent.js");
var import_ColorRule = require("../core/slots/ColorRule.js");
const COMPONENT_NAME = "GaugeChart";
const _GaugeChart = (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    value,
    width,
    height,
    style,
    className,
    prefixIcon,
    seriesActions,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "data-testid": dataTestId = "gauge-chart-root",
    valueAccessor
  } = props;
  const [behavioralTrackingProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(props);
  const config = (0, import_useGaugeChartConfig.useGaugeChartConfig)(props);
  const {
    min,
    max,
    color,
    formatter,
    thresholdsIndicators,
    colorRules,
    toolbar,
    intents,
    tooltip,
    valueAccessor: configValueAccessor,
    loading
  } = config;
  const effectiveValueAccessor = valueAccessor ?? configValueAccessor;
  const chartWidth = (0, import_coerceSizeValue.coerceSizeValue)(width) ?? import_constants.DEFAULT_CHART_WIDTH;
  const chartHeight = (0, import_coerceSizeValue.coerceSizeValue)(height) ?? import_constants.DEFAULT_CHART_HEIGHT;
  const [{ height: containerHeight, width: containerWidth }, containerRef] = (0, import_useObserveContainerSize.useObserveContainerSize)();
  const gaugeChartRef = (0, import_react.useRef)(null);
  (0, import_react.useImperativeHandle)(forwardedRef, () => {
    const consumerConfig = {
      min,
      max,
      color,
      formatter,
      thresholdsIndicators,
      colorRules,
      toolbar,
      tooltip,
      valueAccessor: effectiveValueAccessor,
      loading
    };
    return {
      element: gaugeChartRef.current,
      getConfig: () => JSON.stringify({ type: import_chart_types.ChartType.GaugeChart, ...consumerConfig })
    };
  });
  const ref = (0, import_useMergeRefs.useMergeRefs)([
    forwardedRef,
    gaugeChartRef
  ]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      ...behavioralTrackingProps,
      ref: containerRef,
      style: {
        width: chartWidth,
        height: chartHeight,
        overflow: "hidden",
        ...style
      },
      className,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "data-testid": dataTestId,
      role: "img",
      "aria-label": import_constants.GAUGE_CHART_ARIA_LABEL,
      "data-dt-component": COMPONENT_NAME,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_GaugeChartProviders.GaugeChartProviders,
        {
          config,
          width: containerWidth,
          height: containerHeight,
          value,
          valueAccessor: effectiveValueAccessor,
          explicitColor: color,
          seriesActions,
          loading: loading ?? false,
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_GaugeChartLayout.GaugeChartLayout,
            {
              ref,
              value,
              valueAccessor: effectiveValueAccessor,
              prefixIcon,
              intents,
              toolbar
            }
          )
        }
      )
    }
  );
});
_GaugeChart.displayName = import_constants.DISPLAY_NAME;
const GaugeChart = Object.assign(_GaugeChart, {
  ColorRule: import_ColorRule.ColorRule,
  Intent: import_Intent.Intent,
  Label: import_Label.Label,
  ThresholdIndicator: import_ThresholdIndicator.ThresholdIndicator,
  Toolbar: import_Toolbar.GaugeChartToolbarSlot,
  Tooltip: import_Tooltip.Tooltip
});
