// @dynatrace/strato-components/charts/core/components/axes/components/PanningControls/messages.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var messages_exports = {};
__export(messages_exports, {
  PANNING_CONTROLS_MESSAGES: () => PANNING_CONTROLS_MESSAGES
});
module.exports = __toCommonJS(messages_exports);
var import_react_intl = require("react-intl");
const PANNING_CONTROLS_MESSAGES = (0, import_react_intl.defineMessages)({
  panLeft: {
    id: "1HXxdMWF9JUS/EWC",
    defaultMessage: "Pan left",
    description: "Applies pan to the left"
  },
  panRight: {
    id: "+8N30Kx6cXgWSja4",
    defaultMessage: "Pan right",
    description: "Applies pan to the right"
  },
  buttonLeftLabel: {
    id: "U7gUoyCRKRniBDOp",
    defaultMessage: "Pan to the left",
    description: "Press the button to Pan to the left"
  },
  buttonRightLabel: {
    id: "bvvqbArXvzAMOUit",
    defaultMessage: "Pan to the right",
    description: "Press the button to Pan to the right"
  }
});
