// @dynatrace/strato-components/charts/core/components/axes/components/PanningControls/PanningControls.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var PanningControls_exports = {};
__export(PanningControls_exports, {
  PanningControls: () => PanningControls
});
module.exports = __toCommonJS(PanningControls_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_messages = require("./messages.js");
var import_PanningControls_css = require("./PanningControls.sty.js");
var import_Tooltip = require("../../../../../../overlays/tooltip/Tooltip.js");
const PANNING_STEP = 40;
const PanningControls = (props) => {
  const { showRightControl, showLeftControl, onClickPan } = props;
  const intl = (0, import_react_intl.useIntl)();
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    showLeftControl && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "button",
      {
        "aria-label": intl.formatMessage(
          import_messages.PANNING_CONTROLS_MESSAGES.buttonLeftLabel
        ),
        className: (0, import_PanningControls_css.panningButtons)({ direction: "toLeft" }),
        onClick: () => onClickPan(PANNING_STEP),
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_Tooltip.Tooltip,
          {
            text: intl.formatMessage(import_messages.PANNING_CONTROLS_MESSAGES.panLeft),
            placement: "bottom",
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ArrowSmallLeftIcon, { size: "default", className: import_PanningControls_css.panningArrow })
          }
        )
      }
    ),
    showRightControl && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "button",
      {
        "aria-label": intl.formatMessage(
          import_messages.PANNING_CONTROLS_MESSAGES.buttonRightLabel
        ),
        className: (0, import_PanningControls_css.panningButtons)({ direction: "toRight" }),
        onClick: () => onClickPan(-PANNING_STEP),
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_Tooltip.Tooltip,
          {
            text: intl.formatMessage(import_messages.PANNING_CONTROLS_MESSAGES.panRight),
            placement: "bottom",
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ArrowSmallRightIcon, { size: "default", className: import_PanningControls_css.panningArrow })
          }
        )
      }
    )
  ] });
};
