// @dynatrace/strato-components/charts/core/components/legend/categorical/utils/legend-grid.utils.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var legend_grid_utils_exports = {};
__export(legend_grid_utils_exports, {
  buildGridSettings: () => buildGridSettings,
  getMinMaxSeriesWidth: () => getMinMaxSeriesWidth
});
module.exports = __toCommonJS(legend_grid_utils_exports);
var import_chars_font_width_map = require("../../../text-ellipsis/chars-font-width-map.js");
var import_truncate_string = require("../../../text-ellipsis/truncate-string.js");
var import_constants = require("../constants.js");
const getMinMaxSeriesWidth = (items) => {
  const { shortestSeries, longestSeries } = items.reduce(
    (acc, current) => {
      const labelWidth = (0, import_truncate_string.calculateWidth)(current.name, import_chars_font_width_map.charsFontWidth12pixelsSize500Map) + import_constants.LEGEND_BUTTON_WIDTH;
      return {
        shortestSeries: Math.min(acc.shortestSeries, labelWidth),
        longestSeries: Math.max(acc.longestSeries, labelWidth)
      };
    },
    {
      shortestSeries: Infinity,
      longestSeries: -Infinity
    }
  );
  return {
    shortestSeries,
    longestSeries
  };
};
const buildGridSettings = (items, width, height) => {
  const { shortestSeries, longestSeries } = getMinMaxSeriesWidth(items);
  const maxRowsCanFitTheContainer = Math.floor(height / import_constants.ITEM_HEIGHT.value);
  const maxColumnsCanFitTheContainer = Math.max(
    Math.floor(width / import_constants.MIN_ITEM_WIDTH),
    1
  );
  const minItemWidth = Math.min(
    Math.max(shortestSeries, import_constants.MIN_ITEM_WIDTH),
    width
  );
  const maxItemWidth = Math.min(Math.max(longestSeries, import_constants.MIN_ITEM_WIDTH), width);
  const maxItemsPerRow = Math.floor(width / maxItemWidth);
  const maxNrOfSeriesTruncated = maxItemsPerRow * maxRowsCanFitTheContainer;
  const canFitAllItemsWithoutScrollbar = maxNrOfSeriesTruncated >= items.length;
  return {
    maxRows: maxRowsCanFitTheContainer,
    maxColumns: maxColumnsCanFitTheContainer,
    minItemWidth,
    maxItemWidth,
    canFitAllItemsWithoutScrollbar,
    hasVirtual: !canFitAllItemsWithoutScrollbar && items.length > import_constants.VIRTUALIZED_GRID_THRESHOLD,
    longestSeries
  };
};
