// @dynatrace/strato-components/esm/charts/histogram/HistogramChartRoot.js
import { jsx, jsxs } from "react/jsx-runtime";
import { isEmpty } from "lodash-es";
import { forwardRef } from "react";
import { HistogramChartLegend } from "./components/legend/HistogramChartLegend.js";
import {
  DEFAULT_CHART_HEIGHT,
  DEFAULT_LOADING_STATE
} from "./constants/defaults.js";
import { HistogramChartInternal } from "./HistogramChartInternal.js";
import { useHistogramChartConfig } from "./hooks/useHistogramChartConfig.js";
import { HistogramConfigProvider } from "./providers/HistogramConfigProvider.js";
import { HistogramDataProvider } from "./providers/HistogramDataProvider.js";
import { HistogramFilteredDataProvider } from "./providers/HistogramFilteredDataProvider.js";
import { HistogramStateSyncProvider } from "./providers/HistogramStateSyncProvider.js";
import { StoreProvider } from "./providers/StoreProvider.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { ChartLayout } from "../core/components/chart-layout/ChartLayout.js";
import { useAutoLegendRefresh } from "../core/components/chart-layout/hooks/useAutoLegendRefresh.js";
import { getChartStatesTemplates } from "../core/components/chart-states/utils/get-chart-states-template.js";
import { OverlayProvider } from "../core/components/overlay-tooltip/contexts/OverlayContext.js";
import { TooltipProvider } from "../core/components/overlay-tooltip/contexts/TooltipContext.js";
import { ChartSeriesActionContext } from "../core/components/series-action/ChartSeriesActionContext.js";
import { coerceSizeValue } from "../core/utils/coerceSizeValue.js";
const COMPONENT_NAME_BEHAVIOURAL_TRACKING = "HistogramChart";
const HistogramChartRoot = forwardRef((props, forwardedRef) => {
  const {
    children,
    data,
    height,
    width,
    colorPalette,
    truncationMode,
    loading = DEFAULT_LOADING_STATE,
    seriesActions,
    initialZoom,
    currentZoom,
    onZoomChange,
    infiniteZoom,
    className,
    style,
    valueRepresentation,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow
  } = props;
  const [behaviorTrackingProps] = useBehavioralTrackingProps(props);
  const chartHeight = coerceSizeValue(height ?? DEFAULT_CHART_HEIGHT);
  const chartConfig = useHistogramChartConfig(children, {
    colorPalette,
    truncationMode,
    valueRepresentation,
    initialZoom,
    currentZoom,
    onZoomChange,
    infiniteZoom
  });
  const { emptyState, errorState } = getChartStatesTemplates(props.children);
  const {
    position: legendPosition,
    onRatioChange,
    // eslint-disable-next-line @typescript-eslint/no-deprecated
    ratio,
    initialRatio,
    hidden: isLegendHidden
  } = chartConfig.legend;
  const legendRatio = initialRatio ?? ratio;
  const chartLayoutRef = useAutoLegendRefresh(data);
  return /* @__PURE__ */ jsx(HistogramConfigProvider, { config: chartConfig, children: /* @__PURE__ */ jsx(HistogramDataProvider, { data, children: /* @__PURE__ */ jsx(
    ChartSeriesActionContext.Provider,
    {
      value: {
        // We need a cast here as the provider expects unknown and the
        // react context will not be able to carry the generic anyways.
        chartSeriesActions: seriesActions
      },
      children: /* @__PURE__ */ jsx(StoreProvider, { children: /* @__PURE__ */ jsx(HistogramFilteredDataProvider, { children: /* @__PURE__ */ jsx(
        HistogramStateSyncProvider,
        {
          currentDomain: chartConfig.currentZoom,
          initialDomain: chartConfig.initialZoom,
          infiniteZoom: chartConfig.infiniteZoom,
          children: /* @__PURE__ */ jsx(OverlayProvider, { children: /* @__PURE__ */ jsx(TooltipProvider, { children: /* @__PURE__ */ jsxs(
            ChartLayout,
            {
              ref: chartLayoutRef,
              chartHeight,
              chartWidth: width,
              isEmpty: isEmpty(data),
              emptyState,
              errorState,
              showLoader: loading,
              className,
              style,
              ...behaviorTrackingProps,
              "data-dtrum-mask": dataDtrumMask,
              "data-dtrum-allow": dataDtrumAllow,
              "data-dt-component": COMPONENT_NAME_BEHAVIOURAL_TRACKING,
              children: [
                /* @__PURE__ */ jsx(ChartLayout.Graph, { children: /* @__PURE__ */ jsx(
                  HistogramChartInternal,
                  {
                    ref: forwardedRef,
                    config: chartConfig
                  }
                ) }),
                !isLegendHidden && /* @__PURE__ */ jsx(
                  ChartLayout.Legend,
                  {
                    position: legendPosition,
                    initialRatio: legendRatio,
                    onResize: onRatioChange,
                    children: /* @__PURE__ */ jsx(HistogramChartLegend, {})
                  }
                )
              ]
            }
          ) }) })
        }
      ) }) })
    }
  ) }) });
});
HistogramChartRoot.displayName = "HistogramChartRoot";
export {
  HistogramChartRoot
};
