// @dynatrace/strato-components/esm/charts/histogram/HistogramChart.js
import { jsx } from "react/jsx-runtime";
import { forwardRef } from "react";
import { HistogramChartRoot } from "./HistogramChartRoot.js";
import { Annotations } from "./slots/annotations/Annotations.js";
import { Marker } from "./slots/annotations/Marker.js";
import { AnnotationsTooltip } from "./slots/annotations/Tooltip.js";
import { Track } from "./slots/annotations/Track.js";
import { HistogramChartDownloadCSVSlot } from "./slots/DownloadCSV.js";
import { Legend } from "./slots/Legend.js";
import { Pan } from "./slots/Pan.js";
import { Select } from "./slots/Select.js";
import { ThresholdIndicator } from "./slots/ThresholdIndicator.js";
import { Toolbar } from "./slots/Toolbar.js";
import { Tooltip } from "./slots/Tooltip.js";
import { XAxis } from "./slots/XAxis.js";
import { YAxis } from "./slots/YAxis.js";
import { Zoom } from "./slots/Zoom.js";
import { Intent } from "../../core/slots/Intent/Intent.js";
import { EmptyStateSlot } from "../core/slots/chart-states/EmptyStateSlot.js";
import { ErrorStateSlot } from "../core/slots/chart-states/ErrorStateSlot.js";
import { ColorRule } from "../core/slots/ColorRule.js";
import { isFeatureEnabled } from "../core/utils/is-feature-enabled.js";
import { HistogramToXY } from "../xy-chart-adoption/histogram/HistogramToXY.js";
const HistogramChartSwitcher = forwardRef((props, forwardedRef) => {
  const { __useXYAdoption, ...rest } = props;
  if (__useXYAdoption ?? isFeatureEnabled("histogramToXY")) {
    return /* @__PURE__ */ jsx(HistogramToXY, { ...rest, ref: forwardedRef });
  }
  return /* @__PURE__ */ jsx(HistogramChartRoot, { ...rest, ref: forwardedRef });
});
HistogramChartSwitcher.displayName = "HistogramChart";
const HistogramChart = Object.assign(HistogramChartSwitcher, {
  XAxis,
  YAxis,
  Legend,
  Tooltip,
  EmptyState: EmptyStateSlot,
  ErrorState: ErrorStateSlot,
  Annotations,
  ColorRule,
  ThresholdIndicator,
  Select,
  Intent,
  Toolbar,
  DownloadCSV: HistogramChartDownloadCSVSlot,
  Pan,
  Zoom
});
const HistogramAnnotations = Object.assign(
  {},
  {
    Marker,
    Track,
    Tooltip: AnnotationsTooltip
  }
);
export {
  HistogramAnnotations,
  HistogramChart,
  HistogramChartSwitcher
};
