// @dynatrace/strato-components/esm/charts/honeycomb/constants.js
import Colors from "@dynatrace/strato-design-tokens/colors";
import { SCALE_FALLBACK_COLOR } from "../core/constants.js";
const DISPLAY_NAME = "HoneycombChart";
const DEFAULT_DATASET_TYPE = "numeric";
const DEFAULT_HONEYCOMB_WIDTH = "100%";
const DEFAULT_HONEYCOMB_HEIGHT = 300;
const DEFAULT_HONEYCOMB_SHAPE = "hexagon";
const DEFAULT_HONEYCOMB_TILE_COLOR = SCALE_FALLBACK_COLOR;
const DEFAULT_HONEYCOMB_COLOR_PALETTE = "blue";
const DEFAULT_HONEYCOMB_CATEGORICAL_COLOR_PALETTE = "categorical";
const HONEYCOMB_TILE_STROKE_COLOR = Colors.Theme.Background["10"];
const NON_OVERLAPPING_HEIGHT_HEXS = 0.75;
const NON_OVERLAPPING_HEIGHT_CIRCLES = 0.86;
const NON_OVERLAPPING_HEIGHT_SQUARES = 1;
const NON_OVERLAPPING_HEIGHT = {
  circle: NON_OVERLAPPING_HEIGHT_CIRCLES,
  square: NON_OVERLAPPING_HEIGHT_SQUARES,
  hexagon: NON_OVERLAPPING_HEIGHT_HEXS
};
const NOT_HOVERED_TILE_OPACITY = 0.2;
const HOVERED_TILE_OPACITY = 1;
const DEFAULT_LEGEND_POSITION = "auto";
const DEFAULT_HONEYCOMB_CONFIG = {
  shape: DEFAULT_HONEYCOMB_SHAPE,
  colorScheme: DEFAULT_HONEYCOMB_COLOR_PALETTE,
  toolbar: void 0,
  legend: {},
  formatter: {},
  unit: void 0,
  truncationMode: "middle",
  min: void 0,
  max: void 0,
  showLabels: false,
  colorRules: []
};
const DEFAULT_TILE_NAME = "node";
const initialState = {
  selection: {
    hovered: void 0
  },
  overlay: {
    toolbarVisible: false
  }
};
const bestFitInitialState = {
  hiveHeight: 0,
  tileWidth: 0,
  gap: 0
};
const MAX_GAP_SIZE = 16;
const HONEYCOMB_LABEL_FONT_SIZE = 12;
const HONEYCOMB_VALUE_LABEL_FONT_SIZE = 16;
const HONEYCOMB_LABEL_FONT_WEIGHT = 400;
const HONEYCOMB_VALUE_FONT_WEIGHT = 500;
const HONEYCOMB_LABEL_LINE_HEIGHT = 20;
const HONEYCOMB_VALUE_LINE_HEIGHT = 28;
const HONEYCOMB_DEFAULT_TEXT_ALIGNMENT = "center";
const HONEYCOMB_DEFAULT_MIN_ELLIPSIZED_TEXT = "w...w";
const HONEYCOMB_CHART_ARIA_LABEL = "Honeycomb Chart";
export {
  DEFAULT_DATASET_TYPE,
  DEFAULT_HONEYCOMB_CATEGORICAL_COLOR_PALETTE,
  DEFAULT_HONEYCOMB_COLOR_PALETTE,
  DEFAULT_HONEYCOMB_CONFIG,
  DEFAULT_HONEYCOMB_HEIGHT,
  DEFAULT_HONEYCOMB_SHAPE,
  DEFAULT_HONEYCOMB_TILE_COLOR,
  DEFAULT_HONEYCOMB_WIDTH,
  DEFAULT_LEGEND_POSITION,
  DEFAULT_TILE_NAME,
  DISPLAY_NAME,
  HONEYCOMB_CHART_ARIA_LABEL,
  HONEYCOMB_DEFAULT_MIN_ELLIPSIZED_TEXT,
  HONEYCOMB_DEFAULT_TEXT_ALIGNMENT,
  HONEYCOMB_LABEL_FONT_SIZE,
  HONEYCOMB_LABEL_FONT_WEIGHT,
  HONEYCOMB_LABEL_LINE_HEIGHT,
  HONEYCOMB_TILE_STROKE_COLOR,
  HONEYCOMB_VALUE_FONT_WEIGHT,
  HONEYCOMB_VALUE_LABEL_FONT_SIZE,
  HONEYCOMB_VALUE_LINE_HEIGHT,
  HOVERED_TILE_OPACITY,
  MAX_GAP_SIZE,
  NON_OVERLAPPING_HEIGHT,
  NON_OVERLAPPING_HEIGHT_CIRCLES,
  NON_OVERLAPPING_HEIGHT_HEXS,
  NOT_HOVERED_TILE_OPACITY,
  bestFitInitialState,
  initialState
};
