// @dynatrace/strato-components/esm/charts/honeycomb/HoneycombChart.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import {
  forwardRef,
  useImperativeHandle,
  useRef
} from "react";
import { HoneycombChartLayout } from "./components/HoneycombChartLayout.js";
import {
  DEFAULT_HONEYCOMB_HEIGHT,
  DISPLAY_NAME,
  HONEYCOMB_CHART_ARIA_LABEL
} from "./constants.js";
import { HoneycombColorSchemeContext } from "./context/honeycomb-color-scheme.context.js";
import { HoneycombFormatterContext } from "./context/honeycomb-formatter.context.js";
import { HoneycombLoadingContext } from "./context/honeycomb-loading.context.js";
import { HoneycombShapeContext } from "./context/honeycomb-shape.context.js";
import { HoneycombShowLabelsContext } from "./context/honeycomb-show-labels.context.js";
import { HoneycombTruncationMode } from "./context/honeycomb-truncation-mode.context.js";
import { useHoneycombChartConfig } from "./hooks/use-honeycomb-chart-config.js";
import { HoneycombBaseScaleProvider } from "./providers/honeycomb-base-scale.provider.js";
import { HoneycombConfigurationProvider } from "./providers/honeycomb-configuration.provider.js";
import { HoneycombDatasetTypeProvider } from "./providers/honeycomb-dataset-type.provider.js";
import { HoneycombFilteredDataProvider } from "./providers/honeycomb-filtered-data.provider.js";
import { HoneycombMetadataProvider } from "./providers/honeycomb-metadata.provider.js";
import { HoneycombStoreProvider } from "./providers/honeycomb-store.provider.js";
import { HoneycombChartDownloadCSVSlot } from "./slots/DownloadCSV.js";
import { HoneycombLegend } from "./slots/HoneycombLegend.js";
import { Toolbar } from "./slots/Toolbar.js";
import { Tooltip } from "./slots/Tooltip.js";
import { buildAndDownloadCsv } from "./utils/build-and-download-csv.js";
import { normalizeTilesData } from "./utils/normalize-tiles-data.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useMergeRefs } from "../../core/hooks/useMergeRefs.js";
import { Intent } from "../../core/slots/Intent/Intent.js";
import { defaultChartSize } from "../../core/styles/defaultChartValues.sty.js";
import { ChartSeriesActionContext } from "../core/components/series-action/ChartSeriesActionContext.js";
import { getToolbarConfigWithoutIntents } from "../core/components/toolbar/utils/get-toolbar-config-without-intents.js";
import { EmptyStateSlot } from "../core/slots/chart-states/EmptyStateSlot.js";
import { ErrorStateSlot } from "../core/slots/chart-states/ErrorStateSlot.js";
import { ColorRule } from "../core/slots/ColorRule.js";
import { ChartType } from "../core/types/chart-types.js";
import { coerceSizeValue } from "../core/utils/coerceSizeValue.js";
const COMPONENT_NAME = "HoneycombChart";
const normalizeHoneycombConfig = (config) => {
  const { toolbar: configuredToolbar, ...remainingConfig } = config;
  const toolbar = configuredToolbar ? getToolbarConfigWithoutIntents(configuredToolbar) : void 0;
  return { ...remainingConfig, toolbar };
};
const _HoneycombChart = forwardRef((props, forwardedRef) => {
  const {
    data,
    seriesActions,
    width,
    height,
    style: costumerStyle,
    className: customerClassNames,
    loading,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "data-testid": dataTestId
  } = props;
  const [behavioralTrackingProps] = useBehavioralTrackingProps(props);
  const containerRef = useRef(null);
  const chartWidth = coerceSizeValue(width);
  const chartHeight = coerceSizeValue(height) || DEFAULT_HONEYCOMB_HEIGHT;
  const rootRef = useMergeRefs([
    forwardedRef,
    containerRef
  ]);
  const sanitizedData = normalizeTilesData(data);
  const config = useHoneycombChartConfig(props);
  const formatterContext = {
    formatter: config.formatter,
    unit: config.unit
  };
  useImperativeHandle(forwardedRef, () => {
    const consumerConfig = normalizeHoneycombConfig(config);
    return {
      element: containerRef.current,
      downloadData: () => {
        buildAndDownloadCsv(sanitizedData);
      },
      getConfig: () => {
        return JSON.stringify({
          type: ChartType.Honeycomb,
          ...consumerConfig
        });
      }
    };
  });
  return /* @__PURE__ */ jsx(
    "div",
    {
      ...behavioralTrackingProps,
      ref: rootRef,
      style: { ...costumerStyle, width: chartWidth, height: chartHeight },
      className: clsx(defaultChartSize, customerClassNames),
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "data-dt-component": COMPONENT_NAME,
      "aria-label": HONEYCOMB_CHART_ARIA_LABEL,
      role: "img",
      children: /* @__PURE__ */ jsx(HoneycombLoadingContext.Provider, { value: loading, children: /* @__PURE__ */ jsx(HoneycombConfigurationProvider, { config, children: /* @__PURE__ */ jsx(HoneycombShapeContext.Provider, { value: config.shape, children: /* @__PURE__ */ jsx(HoneycombColorSchemeContext.Provider, { value: config.colorScheme, children: /* @__PURE__ */ jsx(HoneycombTruncationMode.Provider, { value: config.truncationMode, children: /* @__PURE__ */ jsx(HoneycombFormatterContext.Provider, { value: formatterContext, children: /* @__PURE__ */ jsx(
        HoneycombShowLabelsContext.Provider,
        {
          value: config.showLabels,
          children: /* @__PURE__ */ jsx(
            HoneycombFilteredDataProvider,
            {
              data: sanitizedData,
              config,
              children: /* @__PURE__ */ jsx(HoneycombDatasetTypeProvider, { children: /* @__PURE__ */ jsx(HoneycombBaseScaleProvider, { children: /* @__PURE__ */ jsx(
                ChartSeriesActionContext.Provider,
                {
                  value: {
                    // We need a cast here as the provider expects unknown and the
                    // react context will not be able to carry the generic anyways.
                    chartSeriesActions: seriesActions
                  },
                  children: /* @__PURE__ */ jsx(HoneycombStoreProvider, { children: /* @__PURE__ */ jsx(HoneycombMetadataProvider, { children: /* @__PURE__ */ jsx(HoneycombChartLayout, { ...props }) }) })
                }
              ) }) })
            }
          )
        }
      ) }) }) }) }) }) })
    }
  );
});
_HoneycombChart.displayName = DISPLAY_NAME;
const HoneycombChart = Object.assign(_HoneycombChart, {
  EmptyState: EmptyStateSlot,
  ErrorState: ErrorStateSlot,
  Legend: HoneycombLegend,
  Intent,
  ColorRule,
  Toolbar,
  DownloadCSV: HoneycombChartDownloadCSVSlot,
  Tooltip
});
export {
  HoneycombChart,
  _HoneycombChart
};
