// @dynatrace/strato-components/esm/charts/core/components/threshold-bar/indicators/ThresholdIndicator.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { Area, LinePath } from "@visx/shape";
import { useId } from "../../../../../core/hooks/useId.js";
import { ThresholdBrightnessFilter } from "../ThresholdBrightnessFilter.js";
const ThresholdIndicator = (props) => {
  const id = useId();
  const { data, color, isActive = false, strokeOnly = false } = props;
  if (data.length === 0) {
    return null;
  }
  const opacity = isActive ? 0.2 : 0.1;
  const strokeWidth = isActive || strokeOnly ? 1 : 0;
  const strokeDasharray = !isActive && strokeOnly ? "3,2" : "0,0";
  const lineProps = {
    data,
    stroke: color,
    strokeWidth,
    strokeDasharray
  };
  const brightnessFilterId = `threshold-indicator-${id}`;
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(ThresholdBrightnessFilter, { id: brightnessFilterId }),
    /* @__PURE__ */ jsx(
      LinePath,
      {
        ...lineProps,
        role: "graphics-symbol",
        "aria-label": "threshold indicator line 0",
        x: (d) => d.x0,
        y: (d) => d.y0
      }
    ),
    !strokeOnly && /* @__PURE__ */ jsx(
      Area,
      {
        data,
        role: "graphics-symbol",
        "aria-label": "threshold indicator area",
        x0: (d) => d.x0,
        x1: (d) => d.x1,
        y0: (d) => d.y0,
        y1: (d) => d.y1,
        fill: color,
        opacity,
        filter: isActive ? `url(#${brightnessFilterId})` : void 0
      }
    ),
    /* @__PURE__ */ jsx(
      LinePath,
      {
        ...lineProps,
        role: "graphics-symbol",
        "aria-label": "threshold indicator line 1",
        x: (d) => d.x1,
        y: (d) => d.y1
      }
    )
  ] });
};
export {
  ThresholdIndicator
};
