// @dynatrace/strato-components/charts/xy-chart/messages/messages.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var messages_exports = {};
__export(messages_exports, {
  XYCHART_MESSAGES: () => XYCHART_MESSAGES
});
module.exports = __toCommonJS(messages_exports);
var import_react_intl = require("react-intl");
const XYCHART_MESSAGES = {
  separatorDash: (0, import_react_intl.defineMessage)({
    id: "sI/bFgg4u46KiHYZ",
    defaultMessage: "\u2013",
    description: "Text in between the from/to value displayed in the tooltip for the xy chart rect series"
  }),
  zoomModeExplore: (0, import_react_intl.defineMessage)({
    id: "Siq71+wvTgBerJXL",
    defaultMessage: "Explore mode",
    description: 'Text in a tooltip shown when hovering the "Explore" mode in the chart toolbar'
  }),
  zoomModePan: (0, import_react_intl.defineMessage)({
    id: "wjx6xe/Gmj4hLt7b",
    defaultMessage: "Pan mode",
    description: 'Text in a tooltip shown when hovering the "Pan" mode in the chart toolbar'
  }),
  zoomModeZoom: (0, import_react_intl.defineMessage)({
    id: "JNqvsQ757QGyZ+wN",
    defaultMessage: "Zoom mode",
    description: 'Text in a tooltip shown when hovering the "Zoom" mode in the chart toolbar'
  }),
  downloadAsCsv: (0, import_react_intl.defineMessage)({
    id: "B1od+sZZT/rP8Plf",
    defaultMessage: "Download as CSV",
    description: 'Text in a tooltip shown when hovering the "Download data" button in the chart toolbar'
  }),
  zoomIn: (0, import_react_intl.defineMessage)({
    id: "mscZH1VbXmfXn9BL",
    defaultMessage: "Zoom in",
    description: 'Text in a tooltip shown when hovering the "Zoom in" action in the xy chart toolbar'
  }),
  zoomOut: (0, import_react_intl.defineMessage)({
    id: "IjB05bsdexXUZ1sw",
    defaultMessage: "Zoom out",
    description: 'Text in a tooltip shown when hovering the "Zoom out" action in the xy chart toolbar'
  }),
  zoomActionArrowUp: (0, import_react_intl.defineMessage)({
    defaultMessage: "Up",
    id: "DHZv33G8z3rNPFeq",
    description: 'Key shortcut in a tooltip shown when hovering the "Zoom in" action in the xy chart toolbar'
  }),
  zoomActionArrowDown: (0, import_react_intl.defineMessage)({
    defaultMessage: "Down",
    id: "z5mvEP+L6qdzvis2",
    description: 'Key shortcut in a tooltip shown when hovering the "Zoom out" action in xy chart toolbar'
  }),
  resetMode: (0, import_react_intl.defineMessage)({
    id: "zfKEZtJ/4td7IY+1",
    defaultMessage: "Reset",
    description: 'Text in a tooltip shown when hovering the "Reset" action in the xy chart toolbar'
  }),
  to: (0, import_react_intl.defineMessage)({
    id: "dIZIXjFpJvLPz3+g",
    defaultMessage: "to",
    description: "Text in the tooltip header indicating the end of the selected range"
  }),
  datapoints: (0, import_react_intl.defineMessage)({
    id: "1qd7igf1SoPYptNg",
    defaultMessage: "{count} data points",
    description: "Text in the tooltip header showing the number of selected data points"
  }),
  datapoint: (0, import_react_intl.defineMessage)({
    id: "bvs68jg9q7RnuWgF",
    defaultMessage: "{count} data point",
    description: "Text in the tooltip header showing the number of selected data points"
  }),
  noDatapoints: (0, import_react_intl.defineMessage)({
    id: "6GOnX5pixiNcVALE",
    defaultMessage: "No data points",
    description: "Text in the tooltip header showing that are no data points selected"
  }),
  seriesDatapoints: (0, import_react_intl.defineMessage)({
    id: "nDzGVBRq+V8ulNuu",
    defaultMessage: "{count, plural, one {# data point} other {# data points}} across {seriesCount} series",
    description: "Text in the tooltip header showing the number of selected data points in multiple series"
  }),
  categoriesDatapoints: (0, import_react_intl.defineMessage)({
    id: "psQtWNyQryNFymrs",
    defaultMessage: "{count, plural, one {# data point} other {# data points}} in {seriesCount, plural, one {# category} other {# categories}}",
    description: "Text in the tooltip header showing the number of selected data points in multiple categories"
  }),
  zoomToSelection: (0, import_react_intl.defineMessage)({
    id: "CQIPnzlUQ9+DqtjP",
    defaultMessage: "Zoom to selection",
    description: "Text shown in the tooltip as the default action when performing a select interaction in the xy chart"
  }),
  unnamedSeriesMessage: (0, import_react_intl.defineMessage)({
    id: "3oBNy/ExeWqmMWha",
    defaultMessage: "Series {index}",
    description: "Fallback label used for XY chart series when no valid name accessor value is provided."
  })
};
