// @dynatrace/strato-components/esm/charts/xy-chart/components/tooltip/utils/get-custom-tooltip-payload.js
import { getColor } from "./get-color.js";
import { getTooltipFormatter } from "./get-tooltip-formatter.js";
import {
  isAreaSeries,
  isBarSeries,
  isDotSeries,
  isRectSeries
} from "../../../utils/type-guards.js";
const getCustomTooltipPayload = (content, metadata, axisIdToFormatter, intl, colorScaleMap, colorOverride) => {
  const baseTooltipPayload = {
    datapoint: content.datapoint,
    color: getColor(
      content.series,
      content.datapoint,
      metadata,
      colorScaleMap,
      colorOverride
    ),
    seriesId: content.series.id,
    seriesName: metadata.name,
    xFormatter: getTooltipFormatter(metadata.xAxis, axisIdToFormatter, intl),
    yFormatter: getTooltipFormatter(metadata.yAxis, axisIdToFormatter, intl),
    xAxisLabel: content.series.xAxisId,
    yAxisLabel: content.series.yAxisId
  };
  if (isRectSeries(content.series)) {
    const rectMetadata = metadata;
    return {
      rectPayload: {
        ...baseTooltipPayload,
        datapoint: {
          ...baseTooltipPayload.datapoint,
          value: content.datapoint.value
        },
        valueLabel: rectMetadata.valueAccessorLabel ?? rectMetadata.valueAccessor ?? "",
        valueFormatter: rectMetadata.valueFormatter
      }
    };
  } else if (isDotSeries(content.series)) {
    const dotMetadata = metadata;
    return {
      dotPayload: {
        ...baseTooltipPayload,
        shape: dotMetadata.shape
      }
    };
  } else if (isBarSeries(content.series)) {
    return {
      barPayload: {
        ...baseTooltipPayload
      }
    };
  } else if (isAreaSeries(content.series)) {
    return {
      areaPayload: {
        ...baseTooltipPayload
      }
    };
  } else {
    return {
      linePayload: {
        ...baseTooltipPayload
      }
    };
  }
};
export {
  getCustomTooltipPayload
};
