// @dynatrace/strato-components/charts/xy-chart/processing-strategies/utils/cast-data/cast-data-types.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var cast_data_types_exports = {};
__export(cast_data_types_exports, {
  castDataType: () => castDataType,
  castDynamicThresholdsDataType: () => castDynamicThresholdsDataType,
  castStaticThresholdsDataType: () => castStaticThresholdsDataType
});
module.exports = __toCommonJS(cast_data_types_exports);
var import_type_guards = require("../../../../core/types/type-guards.js");
var import_get_y_value = require("../../../components/thresholds/get-y-value.js");
var import_axis = require("../../../types/axis.js");
function isValidDate(d) {
  return d instanceof Date && !isNaN(d.valueOf());
}
const castValueToNumber = {
  Number: (value) => value,
  String: (value) => {
    const castValue = parseFloat(value);
    return isNaN(castValue) ? void 0 : castValue;
  },
  Date: (value) => value.getTime()
};
const castValueToString = {
  Number: (value) => value.toString(),
  String: (value) => value,
  Date: (value) => value.toISOString()
};
const castValueToDate = {
  Number: (value) => {
    const castValue = new Date(value);
    return isValidDate(castValue) ? castValue : void 0;
  },
  String: (value) => {
    const castValue = new Date(value);
    return isValidDate(castValue) ? castValue : void 0;
  },
  Date: (value) => isValidDate(value) ? value : void 0
};
const castsMap = {
  [import_axis.XYAxisTypeLinear]: castValueToNumber,
  [import_axis.XYAxisTypeCategorical]: castValueToString,
  [import_axis.XYAxisTypeTime]: castValueToDate,
  [import_axis.XYAxisTypeLog]: castValueToNumber
};
function castDataType(value, axisType) {
  return castsMap[axisType][value?.constructor.name]?.(
    value
  );
}
function castDynamicThresholdsDataType(data, xAxisType, yAxisType) {
  return data.reduce((acc, d) => {
    const x = castDataType(d.x, xAxisType);
    const yMin = castDataType((0, import_get_y_value.getYValue)(d, "min"), yAxisType);
    const yMax = castDataType((0, import_get_y_value.getYValue)(d, "max"), yAxisType);
    const isValidThreshold = x !== void 0 && yMin !== void 0 && yMax !== void 0;
    if (isValidThreshold) {
      if ((0, import_type_guards.isPointThreshold)(yMin, yMax)) {
        acc.push({ x, y: yMin });
      } else {
        acc.push({ x, yMin, yMax });
      }
    }
    return acc;
  }, []);
}
function castStaticThresholdsDataType(data, yAxisType) {
  const yMin = castDataType((0, import_get_y_value.getYValue)(data, "min"), yAxisType);
  const yMax = castDataType((0, import_get_y_value.getYValue)(data, "max"), yAxisType);
  const isValidThreshold = yMin !== void 0 && yMax !== void 0;
  if (isValidThreshold) {
    if ((0, import_type_guards.isPointThreshold)(yMin, yMax)) {
      return { y: yMin };
    } else {
      return { yMin, yMax };
    }
  }
  return void 0;
}
