// @dynatrace/strato-components/content/markdown/Markdown.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Markdown_exports = {};
__export(Markdown_exports, {
  Markdown: () => Markdown,
  componentMappings: () => componentMappings
});
module.exports = __toCommonJS(Markdown_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_markdown = __toESM(require("react-markdown"));
var import_remark_gfm = __toESM(require("remark-gfm"));
var import_Markdown_css = require("./Markdown.sty.js");
var import_table_mapping = require("./transformers/table-mapping.js");
var import_typography_mapping = require("./transformers/typography-mapping.js");
var import_utils = require("./utils.js");
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useSafeForwardProps = require("../../core/hooks/useSafeForwardProps.js");
var import_textStyle_css = require("../../styles/textStyle.sty.js");
const COMPONENT_NAME = "Markdown";
const componentMappings = {
  ...import_typography_mapping.typographyMappings,
  ...import_table_mapping.tableMappings
};
const Markdown = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    children,
    onSelectionChange,
    customComponentMappings,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const ariaLabelingProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(restProps);
  const forwardProps = (0, import_useSafeForwardProps.useSafeForwardProps)(restProps, ariaLabelingProps);
  const startNode = (0, import_react.useRef)(null);
  const endNode = (0, import_react.useRef)(null);
  const extendedComponentMappings = (0, import_react.useMemo)(() => {
    return customComponentMappings ? {
      ...componentMappings,
      ...customComponentMappings
    } : componentMappings;
  }, [customComponentMappings]);
  const handleSelectionChange = () => {
    const selection = document.getSelection();
    if (!selection || selection.isCollapsed || !selection.anchorNode || !selection.focusNode) {
      return;
    }
    const start = (0, import_utils.findValidParent)(selection.anchorNode);
    const end = (0, import_utils.findValidParent)(selection.focusNode);
    if (!start && !end) {
      return;
    }
    if (start) {
      startNode.current = start;
    }
    if (end) {
      endNode.current = end;
    }
    if (startNode.current && endNode.current) {
      const sourcePos = (0, import_utils.extractSourcePos)(startNode.current, endNode.current);
      if (sourcePos && onSelectionChange) {
        const startOffset = (0, import_utils.findOffset)(sourcePos.start, children, -1);
        const endOffset = (0, import_utils.findOffset)(sourcePos?.end, children);
        sourcePos.start.offset = startOffset;
        sourcePos.end.offset = endOffset;
        onSelectionChange(sourcePos);
      }
    }
  };
  (0, import_react.useEffect)(() => {
    document.addEventListener("selectionchange", handleSelectionChange);
    return () => {
      document.removeEventListener("selectionchange", handleSelectionChange);
    };
  }, [children]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      ref: forwardedRef,
      className: (0, import_clsx.default)(
        (0, import_textStyle_css.textStyle)({ textStyle: "base" }),
        import_Markdown_css.MarkdownWrapperCSS,
        consumerClassName
      ),
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...ariaLabelingProps,
      ...behavioralTrackingProps,
      ...forwardProps,
      "data-dt-component": COMPONENT_NAME,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_react_markdown.default,
        {
          components: extendedComponentMappings,
          remarkPlugins: [import_remark_gfm.default],
          skipHtml: true,
          sourcePos: true,
          children
        }
      )
    }
  );
});
Markdown.displayName = COMPONENT_NAME;
