// @dynatrace/strato-components/content/markdown/utils.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var utils_exports = {};
__export(utils_exports, {
  extractSourcePos: () => extractSourcePos,
  findOffset: () => findOffset,
  findValidParent: () => findValidParent,
  getChildren: () => getChildren,
  matchLanguageParm: () => matchLanguageParm
});
module.exports = __toCommonJS(utils_exports);
var import_lodash_es = require("lodash");
var import_react = require("react");
function hasChildren(element) {
  if ((0, import_react.isValidElement)(element)) {
    return !!element?.props?.children;
  }
  return false;
}
function getChildren(element) {
  return import_react.Children.toArray(hasChildren(element) ? element.props.children : []);
}
function matchLanguageParm(className) {
  return /language-(\w+)/.exec(className || "");
}
function extractSourcePos(start, end) {
  const regexp = /^(\d+):(\d+)-(\d+):(\d+)/;
  const match1 = regexp.exec(
    start.getAttribute("data-sourcepos") ?? ""
  );
  const match2 = regexp.exec(
    end.getAttribute("data-sourcepos") ?? ""
  );
  if (match1 && match2) {
    const [, start1, sOffset1, end1, eOffset1] = match1.filter((m) => !(0, import_lodash_es.isNaN)(m)).map((m) => Number(m));
    const [, start2, sOffset2, end2, eOffset2] = match2.filter((m) => !(0, import_lodash_es.isNaN)(m)).map((m) => Number(m));
    return {
      start: {
        line: start1 < start2 ? start1 : start2,
        column: start1 < start2 ? sOffset1 : sOffset2,
        offset: 0
      },
      end: {
        line: end1 > end2 ? end1 : end2,
        column: end1 > end2 ? eOffset1 : eOffset2,
        offset: 0
      }
    };
  }
  return null;
}
function findValidParent(element, count = 0) {
  if (count > 10 || element === null) {
    return null;
  }
  if (element instanceof Element && element.getAttribute("data-sourcepos")) {
    return element;
  }
  return findValidParent(element.parentNode, count + 1);
}
function findOffset(position, source, lineOffset = 0) {
  let offset = 0;
  const lines = source.split("\n");
  for (let i = 0; i < position.line + lineOffset; i++) {
    if (!(0, import_lodash_es.isNil)(lines[i])) {
      offset += lines[i].length + 1;
    }
  }
  return offset;
}
