// @dynatrace/strato-components/esm/charts/xy-chart/constants/defaults.js
import { initialSelectionState } from "../../core/components/annotations/providers/AnnotationsDataInitialState.js";
import { TRUNCATION_MODE } from "../../core/components/legend/range/constants.js";
import { defaultRatio } from "../../core/constants.js";
const DEFAULT_CHART_HEIGHT = 300;
const DEFAULT_COLOR_PALETTE = "categorical";
const DEFAULT_COLOR_PALETTE_RECT = "blue";
const DEFAULT_LOADING_STATE = false;
const DEFAULT_AXIS_CONFIG = {
  hidden: false,
  min: "auto",
  max: "auto",
  reversed: false
};
const DEFAULT_LEGEND_RANGE_FORMATTER = {
  maximumFractionDigits: 1
};
const initialState = {
  zoom: {
    zoomInEnabled: true,
    zoomOutEnabled: false
  },
  currentDomain: void 0,
  select: void 0,
  interactionMode: "explore",
  thresholds: { hoveredThresholdId: void 0 },
  highlightedSeriesId: void 0,
  filteredSeriesIds: void 0,
  annotations: initialSelectionState
};
const defaultLegend = {
  hidden: false,
  position: "auto",
  ratio: defaultRatio
};
const defaultToolbar = {
  hidden: false,
  placement: "top-right",
  collapsed: false,
  draggable: true
};
const defaultConfig = {
  legend: defaultLegend,
  toolbar: defaultToolbar,
  colorPalette: void 0,
  truncationMode: TRUNCATION_MODE.Middle,
  thresholdIndicators: [],
  yAxes: [],
  xAxes: [],
  colorRules: []
};
const XY_CHART_ARIA_LABEL = "XY Chart";
const ANNOTATIONS_PADDING_BOTTOM = 6;
const DISPLAY_UNITS_LIMIT = 3;
const DEFAULT_POINTS_DISPLAY = "auto";
const RECT_STROKE_WIDTH = 1;
const XY_CHART_BOTTOM_AXIS_LABEL = "Bottom axis";
const XY_CHART_TOP_AXIS_LABEL = "Top axis";
const DISPLAY_NAME = "XYChart";
export {
  ANNOTATIONS_PADDING_BOTTOM,
  DEFAULT_AXIS_CONFIG,
  DEFAULT_CHART_HEIGHT,
  DEFAULT_COLOR_PALETTE,
  DEFAULT_COLOR_PALETTE_RECT,
  DEFAULT_LEGEND_RANGE_FORMATTER,
  DEFAULT_LOADING_STATE,
  DEFAULT_POINTS_DISPLAY,
  DISPLAY_NAME,
  DISPLAY_UNITS_LIMIT,
  RECT_STROKE_WIDTH,
  XY_CHART_ARIA_LABEL,
  XY_CHART_BOTTOM_AXIS_LABEL,
  XY_CHART_TOP_AXIS_LABEL,
  defaultConfig,
  defaultLegend,
  defaultToolbar,
  initialState
};
