// @dynatrace/strato-components/esm/forms/date-time-picker/hooks/useDateTimePicker.js
import { TZDateMini } from "@date-fns/tz";
import { isAfter, isBefore, isValid } from "date-fns";
import { useCallback, useEffect, useMemo, useRef, useState } from "react";
import { getTimezone } from "@dynatrace-sdk/user-preferences";
import { useDateTimeFormat } from "./useDateTimeFormat.js";
import { useSpinButtons } from "./useSpinButtons.js";
import { useAriaLabelingProps } from "../../../core/hooks/useAriaLabelingProps.js";
import { usePrevious } from "../../../core/hooks/usePrevious.js";
import { useFormFieldMessagesContext } from "../../forms-core/hooks/useFormFieldMessagesContext.js";
import {
  getTimeValue,
  removePrecision
} from "../../forms-core/time/timeParser.js";
import { focusElementWithRetry } from "../utils/focusDateTimeInput.js";
function useDateTimePicker(props) {
  const {
    type,
    precision,
    min,
    max,
    disabled = false,
    readOnly = false,
    value: propValue,
    onChange,
    required,
    valueType
  } = props;
  const isDisabled = disabled || valueType !== "iso8601";
  const timezone = getTimezone();
  const { "aria-label": ariaLabelPrefix } = useAriaLabelingProps(props);
  const value = propValue === "" ? null : propValue;
  const [isoDateTime, setIsoDateTime] = useState(
    getTimeValue(value ?? void 0, precision)?.absoluteDate ?? null
  );
  const pendingIsoDateTimeRef = useRef(isoDateTime);
  const { toParts, emptyParts } = useDateTimeFormat(type, precision);
  const {
    state: { isValid: isValidFormControl }
  } = useFormFieldMessagesContext();
  const { spinButtons, dateTime, updateParts, handlers, parts, toDateTime } = useSpinButtons(
    isoDateTime,
    type,
    precision,
    isDisabled,
    readOnly,
    ariaLabelPrefix,
    value
  );
  const prevDateTime = usePrevious(dateTime);
  const minDateTime = useMemo(() => {
    const candidate = min !== void 0 ? new TZDateMini(min, timezone) : void 0;
    return isValid(candidate) ? candidate : void 0;
  }, [min, timezone]);
  const maxDateTime = useMemo(() => {
    const candidate = max !== void 0 ? new TZDateMini(max, timezone) : void 0;
    return isValid(candidate) ? candidate : void 0;
  }, [max, timezone]);
  const applyDateTime = useCallback(
    (isoDateTime2, passFilter) => {
      if (isoDateTime2 === null) {
        updateParts(emptyParts);
        return true;
      }
      const newDate = new TZDateMini(isoDateTime2, timezone);
      newDate.setMinutes(
        newDate.getMinutes() + (type === "date" ? newDate.getTimezoneOffset() : 0)
      );
      if (!isValid(newDate)) {
        return false;
      }
      const newDateParts = toParts(newDate);
      passFilter ? updateParts(newDateParts.filter((p) => passFilter.has(p.type))) : updateParts(newDateParts);
      return true;
    },
    [emptyParts, timezone, toParts, type, updateParts]
  );
  const dateSelectorProps = useMemo(
    () => ({
      value: isoDateTime ?? TZDateMini.tz(timezone).toISOString(),
      onChange: (newIsoString) => {
        applyDateTime(newIsoString, /* @__PURE__ */ new Set(["year", "month", "day"]));
        focusElementWithRetry(() => handlers.get("hour")?.element.current);
      },
      disabled: disabled || readOnly,
      error: false
    }),
    [isoDateTime, timezone, disabled, readOnly, applyDateTime, handlers]
  );
  const error = useMemo(() => {
    if (!isValidFormControl) {
      return true;
    }
    if (!required && isoDateTime === null) {
      return false;
    }
    return (dateTime === null || !isValid(dateTime) || minDateTime && isBefore(dateTime, minDateTime) || maxDateTime && isAfter(dateTime, maxDateTime)) ?? false;
  }, [
    isValidFormControl,
    required,
    isoDateTime,
    dateTime,
    minDateTime,
    maxDateTime
  ]);
  const empty = useMemo(
    () => !parts.some((part) => part.type !== "literal" && part.value !== ""),
    [parts]
  );
  useEffect(() => {
    const numericDateTime = +(dateTime || 0);
    const numericPrevDateTime = +(prevDateTime || 0);
    if (prevDateTime === void 0 || Number.isNaN(numericDateTime) && Number.isNaN(numericPrevDateTime) || numericDateTime === numericPrevDateTime) {
      return;
    }
    if (dateTime === null) {
      pendingIsoDateTimeRef.current = null;
      setIsoDateTime(null);
      onChange?.(null);
      return;
    }
    if (!isValid(dateTime)) {
      pendingIsoDateTimeRef.current = null;
      setIsoDateTime(null);
      onChange?.("Invalid Date");
      return;
    }
    const newIsoDateTime = dateTime.toISOString();
    if (isoDateTime === newIsoDateTime) {
      return;
    }
    pendingIsoDateTimeRef.current = newIsoDateTime;
    setIsoDateTime(newIsoDateTime);
    onChange?.(getTimeValue(newIsoDateTime ?? null, precision) ?? null);
  }, [dateTime, prevDateTime]);
  useEffect(() => {
    if (removePrecision(
      pendingIsoDateTimeRef.current ?? "",
      type === "date" ? "day" : precision
    ) !== value) {
      applyDateTime(
        getTimeValue(value ?? void 0, precision)?.absoluteDate ?? null
      );
    }
  }, [value]);
  return {
    isoDateTime,
    spinButtons,
    handlers,
    dateSelectorProps,
    error,
    empty,
    applyDateTime,
    firstSpinButtonHandler: handlers.get(parts[0]?.type),
    dateTime,
    toDateTime,
    parts
  };
}
export {
  useDateTimePicker
};
