// @dynatrace/strato-components/filters/segment-selector/contexts/SegmentsCache/SegmentsCache.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SegmentsCache_exports = {};
__export(SegmentsCache_exports, {
  getSegment: () => getSegment,
  segmentsClient: () => segmentsClient,
  useSegmentsCache: () => useSegmentsCache
});
module.exports = __toCommonJS(SegmentsCache_exports);
var import_react_query = require("@tanstack/react-query");
var import_app_environment = require("@dynatrace-sdk/app-environment");
var import_client_filter_segment_management = require("@dynatrace-sdk/client-filter-segment-management");
const segmentQueryClient = new import_react_query.QueryClient({
  defaultOptions: {
    queries: {
      refetchOnWindowFocus: false,
      refetchOnMount: true,
      retry: false,
      staleTime: Infinity
    }
  }
});
const currentUser = (0, import_app_environment.getCurrentUserDetails)();
function catchError(error) {
  if (error.name === "403" || error.message === "403") {
    segmentsClient.hasAccessRights = false;
    segmentsClient.queryClient.setDefaultOptions({
      queries: { enabled: false }
    });
  }
}
function getSegmentFromCache(id) {
  const allSegments = segmentsClient.queryClient.getQueryData(["getAllSegments"]);
  const singleSegment = segmentsClient.queryClient.getQueryData([
    "getSegment",
    id
  ]);
  return allSegments?.get(id) ?? singleSegment ?? null;
}
const segmentsClient = {
  /** react-query's queryClient used for segments queries. */
  queryClient: segmentQueryClient,
  /** flag for missing access rights (403) for any segments query. */
  hasAccessRights: void 0
};
function processSegment(segment) {
  const { uid, ...remainingSegment } = segment;
  return {
    segment: { ...remainingSegment, id: uid },
    temporarilyApplied: !(segment.isPublic || segment.owner === currentUser.id)
  };
}
async function getSegment(uid) {
  if (!uid) {
    throw new Error("Segment id is required");
  }
  if (segmentsClient.hasAccessRights === false) {
    throw new Error("403");
  }
  if (segmentsClient.queryClient.isFetching({ queryKey: ["getAllSegments"] }) === 0 && !segmentsClient.queryClient.getQueryState([
    "getAllSegments"
  ])?.isInvalidated) {
    const cachedSegment = getSegmentFromCache(uid);
    if (cachedSegment) {
      return cachedSegment;
    }
  }
  return new Promise((resolve, reject) => {
    import_client_filter_segment_management.filterSegmentsClient.getFilterSegment({
      filterSegmentUid: uid,
      addFields: ["VARIABLES"]
    }).then((storeSegment) => {
      const newCachedSegment = processSegment(storeSegment);
      segmentsClient.hasAccessRights = true;
      resolve(newCachedSegment);
    }).catch((error) => {
      catchError(error);
      reject(error);
      return null;
    });
  });
}
async function getAllSegments() {
  if (segmentsClient.hasAccessRights === false) {
    throw new Error("403");
  }
  return new Promise((resolve, reject) => {
    import_client_filter_segment_management.filterSegmentsClient.getFilterSegments({
      addFields: ["VARIABLES"]
    }).then((segments) => {
      const updatedCache = /* @__PURE__ */ new Map();
      segments.filterSegments.forEach((storeSegment) => {
        const newCachedSegment = processSegment(storeSegment);
        updatedCache.set(storeSegment.uid, newCachedSegment);
        segmentsClient.hasAccessRights = true;
      });
      resolve(updatedCache);
    }).catch((error) => {
      catchError(error);
      reject(error);
    });
  });
}
function useSegmentsCache() {
  return {
    getAllSegments,
    getSegment
  };
}
