// @dynatrace/strato-components/esm/charts/tree-map/components/TreeMapOverlay/TreeMapOverlay.js
import { jsx, jsxs } from "react/jsx-runtime";
import { container } from "./TreeMapOverlay.sty.js";
import { Flex } from "../../../../layouts/flex/Flex.js";
import { useGraphSize } from "../../../core/components/chart-layout/hooks/useGraphSize.js";
import { useTooltipData } from "../../../core/components/overlay-tooltip/_to-be-removed/utils.js";
import { useOverlay } from "../../../core/components/overlay-tooltip/contexts/OverlayContext.js";
import { useSeriesActions } from "../../../core/components/series-action/ChartSeriesActionContext.js";
import { isToolbarEnabled } from "../../../core/utils/is-toolbar-enabled.js";
import { useToolbarConfig } from "../../context/ToolbarContext.js";
import { TreeMapOverlayRefContext } from "../../contexts/TreeMapOverlayRefContext.js";
import { useOverlayHandlers } from "../../hooks/useOverlayHandlers.js";
import { TreeMapToolbar } from "../TreeMapToolbar/TreeMapToolbar.js";
import { TreeMapOverlayTooltip } from "../TreeMapTooltip/TreeMapOverlayTooltip.js";
const TreeMapOverlay = () => {
  const overlay = useOverlay();
  const { width, height } = useGraphSize();
  const roundedWidth = Math.floor(width);
  const roundedHeight = Math.floor(height);
  const {
    handlePointerEnter,
    handlePointerMove,
    handlePointerLeave,
    handleMouseClick
  } = useOverlayHandlers(overlay);
  const toolbarConfig = useToolbarConfig();
  const toolbarEnabled = isToolbarEnabled(toolbarConfig);
  const { chartSeriesActions } = useSeriesActions();
  const data = useTooltipData().data;
  const series = data?.series?.[0];
  const metadata = data?.metadata;
  return /* @__PURE__ */ jsx(TreeMapOverlayRefContext.Provider, { value: overlay.overlayRef, children: /* @__PURE__ */ jsxs(
    Flex,
    {
      ref: overlay.overlayRef,
      style: { width: roundedWidth, height: roundedHeight },
      className: container,
      onClick: handleMouseClick,
      onPointerEnter: handlePointerEnter,
      onPointerMove: handlePointerMove,
      onPointerLeave: handlePointerLeave,
      children: [
        /* @__PURE__ */ jsx(
          TreeMapOverlayTooltip,
          {
            actionsForItem: chartSeriesActions && series && metadata ? () => () => chartSeriesActions({
              id: metadata.id,
              name: series.name,
              value: metadata.value
            }) : void 0
          }
        ),
        toolbarEnabled && /* @__PURE__ */ jsx(TreeMapToolbar, {})
      ]
    }
  ) });
};
TreeMapOverlay["displayName"] = "TreeMapOverlay";
export {
  TreeMapOverlay
};
