// @dynatrace/strato-components/esm/forms/text-area/TextArea.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import {
  forwardRef,
  useImperativeHandle,
  useRef
} from "react";
import { useIntl } from "react-intl";
import { ResizeIcon } from "@dynatrace/strato-icons";
import * as styles from "./TextArea.sty.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useControllableState } from "../../core/hooks/useControllableState.js";
import { useFormControlState } from "../../core/hooks/useFormControlState.js";
import { useFocusRing } from "../../core/styles/useFocusRing.js";
import { mergeProps as _mergeProps } from "../../core/utils/merge-props.js";
import { Flex } from "../../layouts/flex/Flex.js";
import { withFormFieldMessageContextProvider } from "../forms-core/components/form-field/FormFieldMessageContextProvider.js";
import { InternalBaseItem as BaseItem } from "../forms-core/components/form-field-messages/FormFieldMessageBaseItem.js";
import { useFormFieldMessagesContext } from "../forms-core/hooks/useFormFieldMessagesContext.js";
import { useFormFieldValidation } from "../forms-core/hooks/useFormFieldValidation.js";
import { useSyncedFormControlProps } from "../forms-core/hooks/useSyncedFormControlProps.js";
import { FormFieldMessagesAction } from "../forms-core/state/formfield-messages-state-reducer.js";
import { useInputHint } from "../hooks/useInputHint.js";
const COMPONENT_NAME = "TextArea";
const TextAreaRoot = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const intl = useIntl();
  const defaultPlaceholder = intl.formatMessage({
    defaultMessage: "Add text",
    id: "E7kKT3Td1FKB37wr",
    description: "Default placeholder for the text area input."
  });
  const {
    onChange,
    onBlur,
    minLength,
    maxLength,
    id: propId,
    disabled: propDisabled,
    required: propRequired,
    readOnly: propReadOnly,
    resize = "both",
    controlState,
    value: propValue,
    defaultValue,
    width = "full",
    placeholder = defaultPlaceholder,
    "aria-describedby": ariaDescribedby,
    "aria-disabled": ariaDisabledProp,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    className: consumerClassName,
    style: consumerStyle,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
  const { disabled, required, id, ariaDisabled, readOnly } = useSyncedFormControlProps({
    propDisabled,
    propRequired,
    propId,
    propAriaDisabled: ariaDisabledProp,
    propReadOnly
  });
  const disabledOrAriaDisabled = disabled || ariaDisabled || false;
  const { state: messagesContextState, dispatch: messagesContextDispatch } = useFormFieldMessagesContext();
  const containerRef = useRef(null);
  const textareaRef = useRef(null);
  useImperativeHandle(forwardedRef, () => {
    return {
      element: containerRef.current,
      inputRef: textareaRef.current,
      focus() {
        textareaRef.current?.focus();
      },
      validate() {
        textareaRef.current?.checkValidity();
        messagesContextDispatch?.({
          type: FormFieldMessagesAction.SETISTOUCHED,
          payload: { isTouched: true }
        });
      }
    };
  });
  const [controlStateHasError, hint] = useFormControlState(controlState);
  const hasError = controlState ? controlStateHasError : messagesContextState.showError;
  const [value, setValue] = useControllableState({
    defaultValue,
    value: propValue
  });
  const {
    onChange: validationOnChange,
    onBlur: validationOnBlur,
    onInvalid: validationOnInvalid,
    ariaProps
  } = useFormFieldValidation(textareaRef, { required, minLength, maxLength });
  const handleChange = (event) => {
    const newValue = event.target?.value;
    setValue(newValue);
    onChange?.(newValue, event);
    !controlState && validationOnChange();
  };
  const { inputProps, hintProps } = useInputHint({
    ariaDescribedby,
    hint,
    hasError
  });
  const {
    focusProps: { onFocus, onBlur: focusRingOnBlur },
    focusClassName
  } = useFocusRing({
    color: hasError ? "critical" : "neutral",
    ignoreModality: true,
    disabled
  });
  const styleVariants = {
    error: hasError,
    disabled: disabledOrAriaDisabled,
    readOnly
  };
  return /* @__PURE__ */ jsxs(
    Flex,
    {
      gap: 4,
      flexDirection: "column",
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ref: containerRef,
      className: consumerClassName,
      style: consumerStyle,
      ...behavioralTrackingProps,
      "data-dt-component": COMPONENT_NAME,
      children: [
        /* @__PURE__ */ jsxs(
          "div",
          {
            className: styles.container({
              contentWidth: width === "content",
              resize
            }),
            style: { width },
            children: [
              /* @__PURE__ */ jsx(
                "textarea",
                {
                  ref: textareaRef,
                  id,
                  minLength,
                  maxLength,
                  className: clsx(styles.textArea(styleVariants), focusClassName),
                  disabled,
                  required,
                  readOnly,
                  value,
                  "aria-required": required,
                  "aria-disabled": ariaDisabled,
                  placeholder,
                  onChange: handleChange,
                  onInvalid: (e) => !controlState && validationOnInvalid(e),
                  onFocus,
                  onBlur: (e) => {
                    !controlState && validationOnBlur(e);
                    focusRingOnBlur(e);
                    onBlur?.(e);
                  },
                  ..._mergeProps(restProps, controlState ? inputProps : ariaProps)
                }
              ),
              resize !== "none" && /* @__PURE__ */ jsx(ResizeIcon, { className: styles.resizeIcon(styleVariants) })
            ]
          }
        ),
        hint && /* @__PURE__ */ jsx(
          BaseItem,
          {
            variant: hasError ? "error" : "hint",
            disabled: disabled || ariaDisabled,
            ...hintProps,
            children: hint
          }
        )
      ]
    }
  );
});
const TextArea = withFormFieldMessageContextProvider(TextAreaRoot);
TextArea.displayName = COMPONENT_NAME;
export {
  TextArea
};
