// @dynatrace/strato-components/esm/charts/timeseries/utils/convert-to-timeseries/utils/create-timeseries-from-record.js
import { checkTimeseriesFormat } from "./check-timeseries-format.js";
import { extractStartEndDates } from "./extract-start-end-dates.js";
import { extractValue } from "./extractValue.js";
import { getDimensions, getValueDimensions } from "./get-dimensions.js";
import { convertToFieldEntries } from "./getFieldsArrayForRecordIndex.js";
import { createName } from "../create-name.js";
function createTimeseriesFromRecord(record, types) {
  const timeseriesArray = [];
  const timeseriesFields = checkTimeseriesFormat(types.mappings);
  if (!timeseriesFields.isTimeseries) {
    return timeseriesArray;
  }
  const fieldsArray = convertToFieldEntries(types);
  const dimensions = getDimensions(record, fieldsArray, true);
  const [start] = extractStartEndDates(record, fieldsArray);
  const valueDimensions = getValueDimensions(fieldsArray, true);
  for (const valueDimension of valueDimensions ?? []) {
    if (valueDimension.name === "") {
      throw new Error(`Could not calculate key for record ${record}`);
    }
    const timeseries = {
      name: createName(dimensions, valueDimension.name),
      datapoints: [],
      userData: {
        dimensions: [
          ...dimensions,
          {
            field: {
              name: timeseriesFields.durationField,
              type: "duration"
            },
            value: record[timeseriesFields.durationField]
          },
          {
            field: {
              name: timeseriesFields.timeframeField,
              type: "timeframe"
            },
            value: record[timeseriesFields.timeframeField]
          }
        ],
        valueDimension: valueDimension.name
      }
    };
    timeseriesArray.push(timeseries);
    const values = extractValue(
      record,
      valueDimension.name
    );
    const duration = Number(
      BigInt(
        extractValue(record, timeseriesFields.durationField)
      ) / BigInt(1e6)
    );
    if (values !== void 0 && values !== null) {
      let newStart = start;
      for (const value of values) {
        const newEnd = new Date(newStart.getTime() + duration);
        if (value !== null) {
          timeseries.datapoints.push({
            start: newStart,
            end: newEnd,
            value: typeof value === "string" ? parseFloat(value) : value
          });
        }
        newStart = newEnd;
      }
    }
  }
  return timeseriesArray;
}
export {
  createTimeseriesFromRecord
};
