// @dynatrace/strato-components/esm/charts/single-value-grid/utils/build-props-from-grid.js
import { Fragment, jsx } from "react/jsx-runtime";
import { createElement } from "react";
import { Menu } from "../../../navigation/menu/Menu.js";
import { toolbarMenuGroup } from "../../core/components/toolbar/components/ToolbarMenu.sty.js";
import { renderMenuActionChildren } from "../../core/utils/render-menu-action-children.js";
const getIntentsAsyncTemplate = (singleIntents, gridIntents, data, labelText) => async () => {
  const intentsTemplate = gridIntents.length ? [
    /* @__PURE__ */ jsx(Menu.Group, { className: toolbarMenuGroup, children: gridIntents.map((item, key) => /* @__PURE__ */ createElement(Menu.Intent, { ...item, key })) }, "group")
  ] : [];
  if (singleIntents) {
    return (async () => singleIntents(data))().then((template) => {
      return /* @__PURE__ */ jsx(Fragment, { children: template ? [
        /* @__PURE__ */ jsx(
          Menu.Group,
          {
            className: toolbarMenuGroup,
            children: renderMenuActionChildren(template.props.children, {
              divideActionsAndIntents: false
            })
          },
          "individualIntents"
        ),
        gridIntents.length ? /* @__PURE__ */ jsx(Menu.Label, { children: labelText }, "label") : null,
        ...intentsTemplate
      ] : intentsTemplate });
    }).catch(() => /* @__PURE__ */ jsx(Fragment, { children: intentsTemplate }));
  }
  return /* @__PURE__ */ jsx(Fragment, { children: intentsTemplate });
};
const buildPropsFromGrid = (toolbar, singleIntents, gridIntents, labelText) => (data) => {
  if (!singleIntents) {
    return void 0;
  }
  return {
    intentsAsyncTemplate: getIntentsAsyncTemplate(
      singleIntents,
      gridIntents,
      data,
      labelText
    ),
    toolbar: {
      ...toolbar,
      draggable: false
    }
  };
};
export {
  buildPropsFromGrid
};
