// @dynatrace/strato-components/esm/charts/single-value-grid/utils/calculate-grid-layout.js
const GAP = 2;
function calculateGridLayout(containerWidth, containerHeight, childCount) {
  if (containerWidth === 0 || containerHeight === 0 || childCount === 0) {
    return {
      columnCount: 0,
      rowCount: 0,
      lastRowColumnCount: 0,
      gap: "0px"
    };
  }
  const childSize = calculateChildSize(
    containerWidth,
    containerHeight,
    childCount
  );
  const childCountPerRow = Math.floor(containerWidth / childSize);
  const rowCount = Math.ceil(childCount / childCountPerRow);
  const columnCount = Math.ceil(childCount / rowCount);
  const lastRowColumnCount = columnCount - (columnCount * rowCount - childCount);
  return {
    columnCount,
    rowCount,
    lastRowColumnCount,
    gap: `${GAP}px`
  };
}
function calculateChildSize(containerWidth, containerHeight, childCount) {
  const horizontalFitSquareCount = calculatePotentialFittingSquares(
    containerWidth,
    containerHeight,
    childCount
  );
  const verticalFitSquareCount = calculatePotentialFittingSquares(
    containerWidth,
    containerHeight,
    childCount
  );
  const fittingWidth = calculateFittingSideLength(
    containerWidth,
    containerHeight,
    horizontalFitSquareCount,
    childCount
  );
  const fittingHeight = calculateFittingSideLength(
    containerHeight,
    containerWidth,
    verticalFitSquareCount,
    childCount
  );
  return Math.max(fittingWidth, fittingHeight);
}
function calculatePotentialFittingSquares(containerSize1, containerSize2, childCount) {
  return Math.ceil(Math.sqrt(childCount * containerSize1 / containerSize2));
}
function calculateFittingSideLength(containerSize1, containerSize2, potentialFitSquares, childCount) {
  if (Math.floor(potentialFitSquares * containerSize2 / containerSize1) * potentialFitSquares < childCount) {
    return containerSize2 / Math.ceil(potentialFitSquares * containerSize2 / containerSize1);
  } else {
    return containerSize1 / potentialFitSquares;
  }
}
export {
  calculateGridLayout
};
