// @dynatrace/strato-components/charts/core/components/annotations/components/indicators/LineIndicator.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var LineIndicator_exports = {};
__export(LineIndicator_exports, {
  LineIndicator: () => LineIndicator
});
module.exports = __toCommonJS(LineIndicator_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_constants = require("../../constants.js");
const LineIndicator = ({
  x,
  height,
  color,
  filter
}) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
  "line",
  {
    "data-testid": import_constants.TestIds.AnnotationIndicator,
    x1: x,
    x2: x,
    y1: 0,
    y2: height,
    stroke: color,
    strokeWidth: 1,
    filter
  }
);
