// @dynatrace/strato-components/charts/gauge/components/GaugeResponsiveLabel/GaugeResponsiveLabel.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var GaugeResponsiveLabel_exports = {};
__export(GaugeResponsiveLabel_exports, {
  GaugeResponsiveLabel: () => GaugeResponsiveLabel
});
module.exports = __toCommonJS(GaugeResponsiveLabel_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_GaugeResponsiveLabel_css = require("./GaugeResponsiveLabel.sty.js");
var import_Flex = require("../../../../layouts/flex/Flex.js");
var import_useFontSizeForContainer = require("../../../core/hooks/useFontSizeForContainer.js");
var import_useObserveContainerSize = require("../../../core/hooks/useObserveContainerSize.js");
var import_constants = require("../../constants.js");
const GaugeResponsiveLabel = (props) => {
  const { text, prefixIcon, ...paragraphProps } = props;
  const labelRef = (0, import_react.useRef)(null);
  const [{ width, height }, observerRef] = (0, import_useObserveContainerSize.useObserveContainerSize)();
  let preferredFontSize = (0, import_useFontSizeForContainer.useFontSizeForContainer)({
    container: { width, height: height * import_constants.FONT_HEIGHT_RATIO },
    text: prefixIcon ? `00${text}` : `${text}`,
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    element: labelRef.current
  });
  if (preferredFontSize > import_constants.DEFAULT_MAX_VALUE_FONT_SIZE) {
    preferredFontSize = import_constants.DEFAULT_MAX_VALUE_FONT_SIZE;
  }
  const fontSize = `${Math.floor(preferredFontSize)}px`;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Flex.Flex, { className: import_GaugeResponsiveLabel_css.responsiveLabelContainer, ref: observerRef, children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    "span",
    {
      ref: labelRef,
      className: import_GaugeResponsiveLabel_css.parentSvgStyle,
      style: { fontSize, lineHeight: fontSize },
      ...paragraphProps,
      children: [
        prefixIcon,
        text
      ]
    }
  ) });
};
