// @dynatrace/strato-components/esm/charts/categorical-bar/CategoricalBarChartLayout.js
import { jsx, jsxs } from "react/jsx-runtime";
import { forwardRef } from "react";
import { CategoricalBarChart } from "./CategoricalBarChart.js";
import { CategoricalBarChartLegend } from "./components/legend/CategoricalBarChartLegend.js";
import { useCategoricalBarChartBaseConfig } from "./hooks/useCategoricalBarChartBaseConfig.js";
import { useCategoricalBarChartData } from "./hooks/useCategoricalBarChartData.js";
import { isGrouped, isStacked } from "./utils/layout.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { ChartLayout } from "../core/components/chart-layout/ChartLayout.js";
const CategoricalBarChartLayout = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const {
    height,
    width,
    legend,
    isEmpty,
    loading,
    errorState,
    emptyState,
    className,
    style,
    chartLayoutRef,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "data-dt-component": dataDtComponent
  } = props;
  const [behaviorTrackingProps] = useBehavioralTrackingProps(props);
  const { uniqueDimensions } = useCategoricalBarChartData();
  const { groupMode } = useCategoricalBarChartBaseConfig();
  const canShowLegend = isGrouped(groupMode) || isStacked(groupMode) && uniqueDimensions.length > 1;
  const {
    position: legendPosition,
    // eslint-disable-next-line @typescript-eslint/no-deprecated
    ratio,
    initialRatio,
    onRatioChange
  } = legend;
  const legendRatio = initialRatio ?? ratio;
  const showLegend = canShowLegend && !legend.hidden;
  return /* @__PURE__ */ jsxs(
    ChartLayout,
    {
      ref: chartLayoutRef,
      chartHeight: height,
      chartWidth: width,
      isEmpty,
      emptyState,
      errorState,
      showLoader: loading,
      className,
      style,
      ...behaviorTrackingProps,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "data-dt-component": dataDtComponent,
      children: [
        /* @__PURE__ */ jsx(ChartLayout.Graph, { children: /* @__PURE__ */ jsx(CategoricalBarChart, { ref: forwardedRef }) }),
        showLegend && /* @__PURE__ */ jsx(
          ChartLayout.Legend,
          {
            position: legendPosition,
            initialRatio: legendRatio,
            onResize: onRatioChange,
            children: /* @__PURE__ */ jsx(CategoricalBarChartLegend, {})
          }
        )
      ]
    }
  );
});
CategoricalBarChartLayout.displayName = "CategoricalBarChartLayout";
export {
  CategoricalBarChartLayout
};
