// @dynatrace/strato-components/esm/charts/categorical-bar/components/toolbar/CategoricalBarChartToolbar.js
import { jsx } from "react/jsx-runtime";
import { useState } from "react";
import { DownloadDataButton } from "./DownloadDataButton.js";
import { useBehavioralTrackingProps as _useBehavioralTrackingProps } from "../../../../core/hooks/useBehavioralTrackingProps.js";
import { ChartToolbarRenderer } from "../../../core/components/toolbar/ChartToolbarRenderer.js";
import { zIndicesCategoricalAreas } from "../../constants.js";
import { useMouseInBoundsState } from "../../hooks/store/selectors.js";
import { useCategoricalBarChartBaseConfig } from "../../hooks/useCategoricalBarChartBaseConfig.js";
import { useCategoricalBarChartRawData } from "../../hooks/useCategoricalBarChartRawData.js";
import { buildAndDownloadCsv } from "../../utils/build-and-download-csv.js";
const CategoricalBarChartToolbar = ({
  onMouseEnter,
  onMouseLeave
}) => {
  const [isTooltipHidden, setIsTooltipHidden] = useState(false);
  const rawData = useCategoricalBarChartRawData();
  const { toolbar, intents } = useCategoricalBarChartBaseConfig();
  const isInChartBounds = useMouseInBoundsState();
  const isDownloadEnabled = !!toolbar.downloadData?.enabled;
  const behaviorTrackingProps = _useBehavioralTrackingProps(toolbar)[0];
  if (toolbar.hidden || !(isDownloadEnabled || intents.length)) {
    return null;
  }
  const handleToggleMenu = (open) => {
    !open && onMouseLeave();
  };
  const handleDownloadDataToolbar = () => buildAndDownloadCsv(rawData);
  const showDefaultToolbarDownloadButton = isDownloadEnabled && !intents.length;
  const showMenuToolbarButton = isDownloadEnabled && !!intents.length;
  const initialA11yState = showDefaultToolbarDownloadButton ? [{ download: 0 }] : [{ intent: 0 }];
  return /* @__PURE__ */ jsx(
    ChartToolbarRenderer,
    {
      visible: isInChartBounds,
      placement: toolbar.placement,
      collapsed: toolbar.collapsed,
      draggable: true,
      zIndex: zIndicesCategoricalAreas.toolbar,
      initialA11yState,
      onToggleToolbar: setIsTooltipHidden,
      onToggleMenu: handleToggleMenu,
      onMouseEnter,
      onMouseLeave,
      menuDisplayMode: "never",
      intents,
      isDownloadEnabled: showMenuToolbarButton,
      handleDownloadDataToolbar,
      removeDragHandle: true,
      ...behaviorTrackingProps,
      children: showDefaultToolbarDownloadButton && /* @__PURE__ */ jsx(DownloadDataButton, { hideTooltip: isTooltipHidden })
    }
  );
};
export {
  CategoricalBarChartToolbar
};
