// @dynatrace/strato-components/esm/charts/categorical-bar/components/toolbar/DownloadDataButton.js
import { jsx } from "react/jsx-runtime";
import { useRef } from "react";
import { useIntl } from "react-intl";
import { DownloadIcon } from "@dynatrace/strato-icons";
import { CATEGORICAL_MESSAGES } from "./toolbar-tooltip-texts.js";
import { ChartToolbarRenderer } from "../../../core/components/toolbar/ChartToolbarRenderer.js";
import { useA11yToolbarTabIndex } from "../../../core/hooks/useA11yToolbarTabIndex.js";
import { useCategoricalBarChartRawData } from "../../hooks/useCategoricalBarChartRawData.js";
import { buildAndDownloadCsv } from "../../utils/build-and-download-csv.js";
const DownloadDataButton = (props) => {
  const { hideTooltip } = props;
  const downloadDataButtonRef = useRef(null);
  const intl = useIntl();
  const categoricalRawData = useCategoricalBarChartRawData();
  const tabIndex = useA11yToolbarTabIndex("download");
  const handleDownloadDataKeyboard = (event) => {
    if (event.key === "Enter") {
      buildAndDownloadCsv(categoricalRawData);
    }
  };
  const handleDownloadDataToolbar = () => {
    buildAndDownloadCsv(categoricalRawData);
  };
  return /* @__PURE__ */ jsx(
    ChartToolbarRenderer.Control,
    {
      ref: downloadDataButtonRef,
      prefixIcon: /* @__PURE__ */ jsx(DownloadIcon, {}),
      tooltipText: intl.formatMessage(CATEGORICAL_MESSAGES.downloadAsCsv),
      testId: "chart-toolbar-download",
      onKeyDown: handleDownloadDataKeyboard,
      onClick: handleDownloadDataToolbar,
      hideTooltip,
      tabIndex
    }
  );
};
DownloadDataButton["displayName"] = "DownloadDataButton";
export {
  DownloadDataButton
};
