// @dynatrace/strato-geo/map/components/BubbleLayer/BubbleLayer.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var BubbleLayer_exports = {};
__export(BubbleLayer_exports, {
  BubbleLayer: () => BubbleLayer
});
module.exports = __toCommonJS(BubbleLayer_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_maplibre = require("@vis.gl/react-maplibre");
var import_core = require("@dynatrace/strato-components/core");
var import_BubbleCircleLayer = require("./BubbleCircleLayer.js");
var import_BubbleLayerTooltip = require("./BubbleLayerTooltip.js");
var import_BubbleOutlineLayer = require("./BubbleOutlineLayer.js");
var import_size_interpolation_context = require("./contexts/size-interpolation.context.js");
var import_build_radius_scale = require("./utils/build-radius-scale.js");
var import_parse_bubble_data_to_geo_json = require("./utils/parse-bubble-data-to-geo-json.js");
var import_constants = require("../../constants.js");
var import_use_layer_coloring_strategy = require("../../hooks/use-layer-coloring-strategy.js");
var import_use_legend_interaction_highlighting = require("../../hooks/use-legend-interaction-highlighting.js");
var import_use_legend_interaction_visibility = require("../../hooks/use-legend-interaction-visibility.js");
var import_use_resolve_color = require("../../hooks/use-resolve-color.js");
var import_use_tooltip_template = require("../../hooks/use-tooltip-template.js");
var import_BubbleLayerTooltip2 = require("../../slots/BubbleLayerTooltip.js");
var import_bubble_layer = require("../../types/bubble-layer.js");
var import_tooltip_type_guards = require("../../utils/tooltip-type-guards.js");
var import_TooltipWrapper = require("../TooltipWrapper.js");
const BubbleLayer = (props) => {
  const [behavioralTrackingProps] = (0, import_core._useBehavioralTrackingProps)(props);
  const {
    data,
    layerId,
    radius = import_constants.DEFAULT_RADIUS,
    children,
    sizeInterpolation = "fixed"
  } = props;
  const colorParser = (0, import_use_layer_coloring_strategy.useLayerColoringStrategy)();
  const bubbleColor = (0, import_use_resolve_color.useResolveLocationColor)(
    import_constants.DEFAULT_BUBBLE_COLOR,
    props,
    colorParser
  );
  const sourceId = `source-${layerId}`;
  (0, import_use_legend_interaction_highlighting.useLegendInteractionHighlighting)(props, sourceId);
  (0, import_use_legend_interaction_visibility.useLegendInteractionVisibility)(props, layerId);
  let radiusScale = (value) => value;
  if ((0, import_bubble_layer.isScaleRadius)(props)) {
    const { radiusRange = [10, 100], radius: radius2, scale = "linear" } = props;
    radiusScale = (0, import_build_radius_scale.buildRadiusScale)(data, radius2, scale, radiusRange);
  }
  const dataGeoJson = (0, import_parse_bubble_data_to_geo_json.parseBubbleDataToGeoJSON)(
    data,
    bubbleColor,
    radius,
    radiusScale
  );
  const {
    handler: tooltipTemplate,
    hidden: tooltipHidden,
    symbolAlignment
  } = (0, import_use_tooltip_template.useTooltipTemplate)(children, import_BubbleLayerTooltip2.BubbleLayerTooltip, {
    extractProps: true
  });
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_size_interpolation_context.SizeInterpolationContext.Provider, { value: sizeInterpolation, children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_react_maplibre.Source, { id: sourceId, type: "geojson", data: dataGeoJson, generateId: true, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_BubbleCircleLayer.BubbleCircleLayer,
      {
        id: layerId,
        source: sourceId,
        ...behavioralTrackingProps
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_BubbleOutlineLayer.BubbleOutlineLayer, { beforeId: layerId, source: sourceId }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TooltipWrapper.TooltipWrapper, { layerId, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_BubbleLayerTooltip.BubbleLayerTooltip,
      {
        tooltipTemplate: (0, import_tooltip_type_guards.isBubbleTooltipTemplate)(tooltipTemplate) ? tooltipTemplate : void 0,
        hidden: tooltipHidden,
        symbolAlignment
      }
    ) })
  ] }) });
};
BubbleLayer["displayName"] = "BubbleLayer";
