// @dynatrace/strato-components/charts/core/utils/core-scale.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var core_scale_exports = {};
__export(core_scale_exports, {
  buildScale: () => buildScale,
  getBaselineDomain: () => getBaselineDomain,
  getMinMax: () => getMinMax,
  getScale: () => getScale,
  getScaleBoundary: () => getScaleBoundary,
  isDataBoundary: () => isDataBoundary,
  isDefaultBoundary: () => isDefaultBoundary,
  scaleBandInvert: () => scaleBandInvert
});
module.exports = __toCommonJS(core_scale_exports);
var import_d3_scale = require("d3-scale");
var import_lodash_es = require("lodash");
var import_symlog_scale = require("./symlog-scale.js");
function scaleBandInvert(scale) {
  const domain = scale.domain();
  const paddingOuter = scale(domain[0]) ?? 0;
  const numberOfBands = scale.step();
  return function(mousePosition) {
    const halfInnerPadding = scale.step() * scale.paddingInner() / 2;
    const index = Math.floor(
      (mousePosition + halfInnerPadding - paddingOuter) / numberOfBands
    );
    return domain[Math.max(0, Math.min(index, domain.length - 1))];
  };
}
const buildScale = (scaleType, range, dataBoundaries, configMinMax) => {
  const minMaxOptions = {
    niceMin: isDefaultBoundary(configMinMax?.min),
    niceMax: isDefaultBoundary(configMinMax?.max)
  };
  const { min, max } = getMinMax(
    scaleType,
    dataBoundaries,
    configMinMax,
    minMaxOptions
  );
  return getScale(scaleType, [min, max], range);
};
const getScale = (scaleType, domain, range = [0, 1]) => (scaleType === "linear" ? (0, import_d3_scale.scaleLinear)() : (0, import_symlog_scale.getSymlogScale)()).domain(domain).range(range);
const MIN_MAX_DEFAULT_OPTIONS = {
  niceMin: true,
  niceMax: true
};
const getMinMax = (scaleType, dataBoundaries, configMinMax, options = MIN_MAX_DEFAULT_OPTIONS) => {
  const baselineDomain = getBaselineDomain(scaleType, dataBoundaries);
  let min = getScaleBoundary(
    dataBoundaries.min,
    baselineDomain.min,
    configMinMax?.min
  );
  let max = getScaleBoundary(
    dataBoundaries.max,
    baselineDomain.max,
    configMinMax?.max
  );
  const niceScale = getScale(scaleType, [min, max]).nice().domain();
  const scale = getScale(scaleType, [min, max]).domain();
  const [defaultMin, defaultMax] = scale;
  const [niceMin, niceMax] = niceScale;
  min = defaultMin;
  max = defaultMax;
  if (options.niceMin) {
    min = niceMin;
  }
  if (options.niceMax) {
    max = niceMax;
  }
  return { min, max };
};
const getBaselineDomain = (scaleType, dataBoundaries) => {
  let [baselineMin, baselineMax] = [0, 0];
  if (scaleType === "log") {
    baselineMin = 1;
    baselineMax = -1;
  }
  const baselineDomainMin = dataBoundaries.min > baselineMin ? baselineMin : dataBoundaries.min;
  const baselineDomainMax = dataBoundaries.max < baselineMax ? baselineMax : dataBoundaries.max;
  return { min: baselineDomainMin, max: baselineDomainMax };
};
const getScaleBoundary = (dataBoundary, baselineBoundary, scaleBoundary) => {
  if ((0, import_lodash_es.isNumber)(scaleBoundary)) {
    return scaleBoundary;
  }
  if (isDefaultBoundary(scaleBoundary)) {
    return baselineBoundary;
  }
  return dataBoundary;
};
const isDefaultBoundary = (boundary) => (0, import_lodash_es.isUndefined)(boundary) || boundary === "auto";
const isDataBoundary = (boundary) => !(0, import_lodash_es.isUndefined)(boundary) && (boundary === "data-min" || boundary === "data-max");
