// @dynatrace/strato-components/charts/core/utils/format-value.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var format_value_exports = {};
__export(format_value_exports, {
  DEFAULT_PRECISION: () => DEFAULT_PRECISION,
  MAX_PRECISION_ALLOWED: () => MAX_PRECISION_ALLOWED,
  defaultTickValueFormatter: () => defaultTickValueFormatter,
  formatValue: () => formatValue,
  getFormatterPrecision: () => getFormatterPrecision,
  getTickValues: () => getTickValues,
  tickValueFormatter: () => tickValueFormatter
});
module.exports = __toCommonJS(format_value_exports);
var import_lodash_es = require("lodash");
var import_units = require("@dynatrace-sdk/units");
var import_format_date_ticks = require("./format-date-ticks.js");
var import_format_value_by_type = require("./format-value-by-type.js");
const DEFAULT_PRECISION = 1;
const MAX_PRECISION_ALLOWED = 20;
const formatValue = (value, unit = "", precision = DEFAULT_PRECISION) => {
  if (typeof unit === "string") {
    return (0, import_units.format)(value, {
      ...(0, import_format_value_by_type.getFormatterDefinition)(unit),
      maximumFractionDigits: precision
    });
  }
  return (0, import_units.format)(value, { input: unit, maximumFractionDigits: precision });
};
function getFormatterPrecision(tickValues, formatter, unit) {
  for (let minPrecision = DEFAULT_PRECISION; minPrecision <= MAX_PRECISION_ALLOWED; minPrecision++) {
    const formattedTicks = tickValues.map(
      (value) => formatter(value, unit, minPrecision)
    );
    if (allUniq(formattedTicks)) {
      return minPrecision;
    }
  }
  return MAX_PRECISION_ALLOWED;
}
const allUniq = (arr) => (0, import_lodash_es.uniq)(arr).length === arr.length;
const defaultTickValueFormatter = (value, unit) => {
  if (typeof value === "string") {
    return `${value}${(0, import_lodash_es.isEmpty)(unit) ? "" : ` ${unit}`}`;
  }
  if (value instanceof Date) {
    return (0, import_format_date_ticks.formatDateTicks)(value);
  }
  return formatValue(value, unit, DEFAULT_PRECISION);
};
const tickValueFormatter = (tickFormatter, unit) => (value) => tickFormatter && typeof value === "number" ? tickFormatter(value, unit) : defaultTickValueFormatter(value, unit);
const getTickValues = (originalTicks, formatter) => originalTicks.map((value) => formatter?.(value) ?? value.toString());
