// @dynatrace/strato-components/charts/core/utils/format-dates.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var format_dates_exports = {};
__export(format_dates_exports, {
  formatDay: () => formatDay,
  formatHour: () => formatHour,
  formatMinute: () => formatMinute,
  formatMonth: () => formatMonth,
  formatSeconds: () => formatSeconds,
  formatWeekday: () => formatWeekday,
  formatYear: () => formatYear
});
module.exports = __toCommonJS(format_dates_exports);
var import_units = require("@dynatrace-sdk/units");
const formatSeconds = (date) => (0, import_units.formatDate)(date, {
  hour: "2-digit",
  minute: "2-digit",
  second: "2-digit",
  fractionalSecondDigits: 3
});
const formatMinute = (date) => (0, import_units.formatDate)(date, {
  hour: "2-digit",
  minute: "2-digit",
  second: "2-digit"
});
const formatHour = (date) => (0, import_units.formatDate)(date, {
  hour: "2-digit",
  minute: "2-digit"
});
const formatDay = (date) => (0, import_units.formatDate)(date, {
  hour: "2-digit"
});
const formatWeekday = (date) => (0, import_units.formatDate)(date, {
  month: "short",
  day: "numeric"
});
const formatMonth = (date) => (0, import_units.formatDate)(date, {
  month: "short",
  day: "numeric"
});
const formatYear = (date) => (0, import_units.formatDate)(date, {
  year: "numeric"
});
