// @dynatrace/strato-components/charts/core/utils/get-data-point-is-gap.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var get_data_point_is_gap_exports = {};
__export(get_data_point_is_gap_exports, {
  getDatapointIsGap: () => getDatapointIsGap
});
module.exports = __toCommonJS(get_data_point_is_gap_exports);
function getDatapointIsGap(datapoint, nextDatapoint, threshold = 0) {
  return hasGap(datapoint, nextDatapoint, threshold);
}
function hasGap(currentDatapoint, nextDatapoint, threshold) {
  if (!currentDatapoint || !nextDatapoint) {
    return false;
  }
  const distance = Math.abs(getDistance(nextDatapoint, currentDatapoint));
  const hasGap2 = distance > 0;
  const isGapOutsideThreshold = threshold < distance;
  return hasGap2 && isGapOutsideThreshold;
}
function isNumericalDatapoint(datapoint) {
  return typeof datapoint.start === "number";
}
function isTimeframeDatapoint(datapoint) {
  return datapoint.start instanceof Date;
}
function getDistance(nextDatapoint, currentDatapoint) {
  if (isNumericalDatapoint(nextDatapoint) && isNumericalDatapoint(currentDatapoint)) {
    return nextDatapoint.start - currentDatapoint.start;
  }
  if (isTimeframeDatapoint(nextDatapoint) && isTimeframeDatapoint(currentDatapoint)) {
    return nextDatapoint.start.getTime() - resolveEndDate(currentDatapoint).getTime();
  }
  return 0;
}
function resolveEndDate(datapoint) {
  return datapoint.end ?? datapoint.start;
}
