// @dynatrace/strato-components/charts/core/utils/text-truncation.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var text_truncation_exports = {};
__export(text_truncation_exports, {
  ELLIPSIS: () => ELLIPSIS,
  TRUNCATE_LABEL: () => TRUNCATE_LABEL,
  endTruncation: () => endTruncation,
  middleTruncation: () => middleTruncation,
  startTruncation: () => startTruncation,
  truncateText: () => truncateText
});
module.exports = __toCommonJS(text_truncation_exports);
const truncateText = (text, truncationMode, stopTruncation) => {
  let truncatedText = text;
  while (!stopTruncation(truncatedText)) {
    const filteredText = truncatedText.replace(/\.\.\./, "");
    if (!filteredText.length) {
      break;
    }
    truncatedText = TRUNCATE_LABEL[truncationMode](filteredText);
  }
  return truncatedText;
};
const ELLIPSIS = "...";
const middleTruncation = (text) => {
  const middle = Math.floor(text.length / 2);
  const firstHalf = text.slice(0, middle);
  const secondHalf = text.slice(middle + 1);
  const hasEllipsis = firstHalf.length || secondHalf.length;
  const newText = `${firstHalf}${hasEllipsis ? ELLIPSIS : ""}${secondHalf}`;
  return newText;
};
const startTruncation = (text) => {
  const slicedText = text.slice(1);
  const hasEllipsis = slicedText.length;
  const newText = `${hasEllipsis ? ELLIPSIS : ""}${slicedText}`;
  return newText;
};
const endTruncation = (text) => {
  const slicedText = text.slice(0, -1);
  const hasEllipsis = slicedText.length;
  const newText = `${text.slice(0, -1)}${hasEllipsis ? ELLIPSIS : ""}`;
  return newText;
};
const TRUNCATE_LABEL = {
  start: startTruncation,
  end: endTruncation,
  middle: middleTruncation
};
