// @dynatrace/strato-components/charts/core/utils/threshold-to-marker.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var threshold_to_marker_exports = {};
__export(threshold_to_marker_exports, {
  transformThresholdsToMarkers: () => transformThresholdsToMarkers
});
module.exports = __toCommonJS(threshold_to_marker_exports);
var import_d3_scale = require("d3-scale");
var import_lodash_es = require("lodash");
var import_detect_scale = require("./detect-scale.js");
var import_get_marker_value = require("../components/threshold-bar/utils/get-marker-value.js");
var import_get_threshold_pixel_ts = require("../components/threshold-bar/utils/get-threshold-pixel.ts.js");
var import_messages = require("../messages.js");
var import_type_guards = require("../types/type-guards.js");
function findMinValue(thresholds) {
  return thresholds.reduce(
    (min, current) => current.value < min.value ? current : min,
    thresholds[0]
  );
}
function findMaxValue(thresholds) {
  return thresholds.reduce(
    (max, current) => current.value > max.value ? current : max,
    thresholds[0]
  );
}
function findMinRange(thresholds) {
  return thresholds.reduce(
    (min, current) => current.min < min.min ? current : min,
    thresholds[0]
  );
}
function findMaxRange(thresholds) {
  return thresholds.reduce(
    (max, current) => current.max > max.max ? current : max,
    thresholds[0]
  );
}
const isDataRange = (obj) => {
  return typeof obj === "object" && obj !== null && "min" in obj && "max" in obj;
};
const getYValue = (data, type) => {
  if (isDataRange(data)) {
    return data[type];
  }
  return data.value;
};
function findPointWithMinX(data) {
  return (0, import_lodash_es.minBy)(
    data,
    (threshold) => ("point" in threshold ? threshold.point : threshold.date)?.valueOf()
  ) ?? data[0];
}
function findPointWithMaxX(data) {
  return (0, import_lodash_es.maxBy)(
    data,
    (threshold) => ("point" in threshold ? threshold.point : threshold.date)?.valueOf()
  ) ?? data[data.length - 1];
}
function getYRangeFromPoint(threshold) {
  return {
    y0: getYValue(threshold, "min"),
    y1: getYValue(threshold, "max")
  };
}
function interpolateYRange(data, xScale, targetX) {
  const xValues = data.map(
    (threshold) => Number("point" in threshold ? threshold.point : threshold.date)
  );
  const minYValues = data.map((threshold) => getYValue(threshold, "min"));
  const maxYValues = data.map((threshold) => getYValue(threshold, "max"));
  const minYScale = (0, import_detect_scale.isSymlogScale)(xScale) ? (0, import_d3_scale.scaleLog)(xValues, minYValues).clamp(true) : (0, import_d3_scale.scaleLinear)(xValues, minYValues).clamp(true);
  const maxYScale = (0, import_detect_scale.isSymlogScale)(xScale) ? (0, import_d3_scale.scaleLog)(xValues, maxYValues).clamp(true) : (0, import_d3_scale.scaleLinear)(xValues, maxYValues).clamp(true);
  return {
    y0: minYScale(targetX),
    y1: maxYScale(targetX)
  };
}
function getDynamicThresholdDomain(data, side, visibleDomain, xScale) {
  const minXPoint = findPointWithMinX(data);
  const maxXPoint = findPointWithMaxX(data);
  const targetXDomain = side === "left" ? visibleDomain.min : visibleDomain.max;
  const minPointValue = ("point" in minXPoint ? minXPoint.point : minXPoint.date)?.valueOf() ?? 0;
  const maxPointValue = ("point" in maxXPoint ? maxXPoint.point : maxXPoint.date)?.valueOf() ?? 0;
  if (targetXDomain <= minPointValue) {
    return getYRangeFromPoint(minXPoint);
  }
  if (targetXDomain >= maxPointValue) {
    return getYRangeFromPoint(maxXPoint);
  }
  return interpolateYRange(data, xScale, targetXDomain);
}
const buildShadowMarker = (threshold, yScale) => {
  let minThreshold;
  let maxThreshold;
  if ((0, import_type_guards.isDynamic)(threshold.data) || (0, import_type_guards.isDynamicNumeric)(threshold.data)) {
    minThreshold = findMinValue(threshold.data);
    maxThreshold = findMaxValue(threshold.data);
  } else {
    minThreshold = findMinRange(
      threshold.data
    );
    maxThreshold = findMaxRange(
      threshold.data
    );
  }
  return {
    ...threshold,
    y0: (0, import_get_threshold_pixel_ts.getThresholdPixel)(getYValue(minThreshold, "min"), yScale),
    y1: (0, import_get_threshold_pixel_ts.getThresholdPixel)(getYValue(maxThreshold, "max"), yScale),
    value: import_messages.THRESHOLD.dynamic.defaultMessage,
    isShadow: true
  };
};
const buildMarker = (threshold, thresholdBarPosition, xScale, yScale) => {
  const domain = xScale.domain();
  const xMinMax = {
    min: domain[0].valueOf(),
    max: domain[1].valueOf()
  };
  const { y0, y1 } = getDynamicThresholdDomain(
    threshold.data,
    thresholdBarPosition,
    xMinMax,
    xScale
  );
  return {
    ...threshold,
    y0: (0, import_get_threshold_pixel_ts.getThresholdPixel)(y0, yScale),
    y1: (0, import_get_threshold_pixel_ts.getThresholdPixel)(y1, yScale),
    value: import_messages.THRESHOLD.dynamic.defaultMessage
  };
};
const transformThresholdsToMarkers = (thresholdBarPosition, xScale, yScale, formatter) => (acc, threshold) => {
  if (thresholdBarPosition === threshold.position) {
    if ((0, import_type_guards.isStatic)(threshold.data) || (0, import_type_guards.isStaticRange)(threshold.data)) {
      const y0Value = getYValue(threshold.data, "min");
      const y1Value = getYValue(threshold.data, "max");
      acc.push({
        ...threshold,
        y0: (0, import_get_threshold_pixel_ts.getThresholdPixel)(y0Value, yScale),
        y1: (0, import_get_threshold_pixel_ts.getThresholdPixel)(y1Value, yScale),
        value: (0, import_get_marker_value.getMarkerValue)({ y0: y0Value, y1: y1Value }, formatter)
      });
    } else {
      const shadowMarker = buildShadowMarker(threshold, yScale);
      acc.push(shadowMarker);
      const marker = buildMarker(
        threshold,
        thresholdBarPosition,
        xScale,
        yScale
      );
      acc.push(marker);
    }
  }
  return acc;
};
