// @dynatrace/strato-components/charts/core/utils/thresholds-to-indicator.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var thresholds_to_indicator_exports = {};
__export(thresholds_to_indicator_exports, {
  transformHistogramThresholdIntoIndicator: () => transformHistogramThresholdIntoIndicator,
  transformThresholdIntoIndicator: () => transformThresholdIntoIndicator
});
module.exports = __toCommonJS(thresholds_to_indicator_exports);
var import_get_threshold_pixel_ts = require("../components/threshold-bar/utils/get-threshold-pixel.ts.js");
var import_type_guards = require("../types/type-guards.js");
const transformStatic = (data, yScale, chartWidth) => {
  const yScaleValue = (0, import_get_threshold_pixel_ts.getThresholdPixel)(data.value, yScale);
  return [
    {
      x0: 0,
      x1: 0,
      y0: yScaleValue,
      y1: yScaleValue
    },
    {
      x0: chartWidth,
      x1: chartWidth,
      y0: yScaleValue,
      y1: yScaleValue
    }
  ];
};
const transformStaticRange = (data, yScale, chartWidth) => {
  const yScaleMinValue = (0, import_get_threshold_pixel_ts.getThresholdPixel)(data.min, yScale);
  const yScaleMaxValue = (0, import_get_threshold_pixel_ts.getThresholdPixel)(data.max, yScale);
  return [
    {
      x0: 0,
      x1: 0,
      y0: yScaleMinValue,
      y1: yScaleMaxValue
    },
    {
      x0: chartWidth,
      x1: chartWidth,
      y0: yScaleMinValue,
      y1: yScaleMaxValue
    }
  ];
};
const transformDynamic = (data, xScale, yScale) => {
  const orderedData = data.sort(
    (d1, d2) => d1.date.getTime() - d2.date.getTime()
  );
  const [xMin, xMax] = xScale.domain();
  const [firstItem] = orderedData;
  if (xMin < firstItem.date) {
    orderedData.unshift({ value: firstItem.value, date: xMin });
  }
  const lastItem = orderedData[orderedData.length - 1];
  if (xMax > lastItem.date) {
    orderedData.push({
      value: lastItem.value,
      date: xMax
    });
  }
  return data.map((thresholdData) => ({
    x0: xScale(thresholdData.date) ?? 0,
    x1: xScale(thresholdData.date) ?? 0,
    y0: (0, import_get_threshold_pixel_ts.getThresholdPixel)(thresholdData.value, yScale),
    y1: (0, import_get_threshold_pixel_ts.getThresholdPixel)(thresholdData.value, yScale)
  }));
};
const transformDynamicNumeric = (data, xScale, yScale) => {
  const orderedData = data.sort((d1, d2) => d1.point - d2.point);
  const [xMinDate, xMaxDate] = xScale.domain();
  const [firstItem] = orderedData;
  const xMin = xMinDate.valueOf();
  const xMax = xMaxDate.valueOf();
  if (xMin < firstItem.point) {
    orderedData.unshift({ value: firstItem.value, point: +xMin });
  }
  const lastItem = orderedData[orderedData.length - 1];
  if (xMax > lastItem.point) {
    orderedData.push({
      value: lastItem.value,
      point: xMax
    });
  }
  return data.map((thresholdData) => ({
    x0: xScale(thresholdData.point) ?? 0,
    x1: xScale(thresholdData.point) ?? 0,
    y0: (0, import_get_threshold_pixel_ts.getThresholdPixel)(thresholdData.value, yScale),
    y1: (0, import_get_threshold_pixel_ts.getThresholdPixel)(thresholdData.value, yScale)
  }));
};
const transformDynamicRange = (data, xScale, yScale) => {
  const orderedData = data.sort(
    (d1, d2) => d1.date.getTime() - d2.date.getTime()
  );
  const [xMin, xMax] = xScale.domain();
  const [firstItem] = orderedData;
  if (xMin < firstItem.date) {
    orderedData.unshift({
      min: firstItem.min,
      max: firstItem.max,
      date: xMin
    });
  }
  const lastItem = orderedData[orderedData.length - 1];
  if (xMax > lastItem.date) {
    orderedData.push({
      min: lastItem.min,
      max: lastItem.max,
      date: xMax
    });
  }
  return data.map((thresholdData) => ({
    x0: xScale(thresholdData.date) ?? 0,
    x1: xScale(thresholdData.date) ?? 0,
    y0: (0, import_get_threshold_pixel_ts.getThresholdPixel)(thresholdData.min, yScale),
    y1: (0, import_get_threshold_pixel_ts.getThresholdPixel)(thresholdData.max, yScale)
  }));
};
const transformDynamicNumericRange = (data, xScale, yScale) => {
  const orderedData = data.sort((d1, d2) => d1.point - d2.point);
  const [xMinDate, xMaxDate] = xScale.domain();
  const [firstItem] = orderedData;
  const xMin = xMinDate.valueOf();
  const xMax = xMaxDate.valueOf();
  if (xMin < firstItem.point) {
    orderedData.unshift({
      min: firstItem.min,
      max: firstItem.max,
      point: xMin
    });
  }
  const lastItem = orderedData[orderedData.length - 1];
  if (xMax > lastItem.point) {
    orderedData.push({
      min: lastItem.min,
      max: lastItem.max,
      point: xMax
    });
  }
  return data.map((thresholdData) => ({
    x0: xScale(thresholdData.point) ?? 0,
    x1: xScale(thresholdData.point) ?? 0,
    y0: (0, import_get_threshold_pixel_ts.getThresholdPixel)(thresholdData.min, yScale),
    y1: (0, import_get_threshold_pixel_ts.getThresholdPixel)(thresholdData.max, yScale)
  }));
};
const transformThresholdIntoIndicator = (data, chartWidth, yScale, xScale) => {
  if ((0, import_type_guards.isStatic)(data)) {
    return transformStatic(data, yScale, chartWidth);
  } else if ((0, import_type_guards.isStaticRange)(data)) {
    return transformStaticRange(data, yScale, chartWidth);
  } else if ((0, import_type_guards.isDynamic)(data)) {
    return transformDynamic(data, xScale, yScale);
  } else if ((0, import_type_guards.isDynamicRange)(data)) {
    return transformDynamicRange(data, xScale, yScale);
  }
  return [];
};
const transformHistogramThresholdIntoIndicator = (data, chartWidth, yScale, xScale) => {
  if ((0, import_type_guards.isStatic)(data)) {
    return transformStatic(data, yScale, chartWidth);
  } else if ((0, import_type_guards.isStaticRange)(data)) {
    return transformStaticRange(data, yScale, chartWidth);
  } else if ((0, import_type_guards.isDynamicNumeric)(data)) {
    return transformDynamicNumeric(data, xScale, yScale);
  } else if ((0, import_type_guards.isDynamicNumericRange)(data)) {
    return transformDynamicNumericRange(data, xScale, yScale);
  }
  return [];
};
