// @dynatrace/strato-components/charts/core/utils/ticks.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ticks_exports = {};
__export(ticks_exports, {
  DEFAULT_TICK_HEIGHT: () => DEFAULT_TICK_HEIGHT,
  DEFAULT_TICK_WIDTH: () => DEFAULT_TICK_WIDTH,
  DEFAULT_TICK_WIDTH_GANTT: () => DEFAULT_TICK_WIDTH_GANTT,
  getGanttHorizontalTimeAxisNumTicks: () => getGanttHorizontalTimeAxisNumTicks,
  getHorizontalNumTicks: () => getHorizontalNumTicks,
  getHorizontalTimeAxisNumTicks: () => getHorizontalTimeAxisNumTicks,
  getVerticalNumTicks: () => getVerticalNumTicks
});
module.exports = __toCommonJS(ticks_exports);
const DEFAULT_TICK_HEIGHT = 50;
const DEFAULT_TICK_WIDTH = 150;
const DEFAULT_TICK_WIDTH_GANTT = 200;
function getVerticalNumTicks(height) {
  return Math.ceil(height / DEFAULT_TICK_HEIGHT);
}
function getHorizontalNumTicks(width) {
  return Math.ceil(width / DEFAULT_TICK_WIDTH);
}
function getHorizontalTimeAxisNumTicks(width) {
  return Math.min(Math.max(Math.round(width / 130), 2), 20);
}
function getGanttHorizontalTimeAxisNumTicks(width) {
  return Math.floor(width / DEFAULT_TICK_WIDTH_GANTT) || 1;
}
