// @dynatrace/strato-components/charts/core/utils/timeThreshold.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var timeThreshold_exports = {};
__export(timeThreshold_exports, {
  convertTimeThreshold: () => convertTimeThreshold,
  parseTimeThreshold: () => parseTimeThreshold
});
module.exports = __toCommonJS(timeThreshold_exports);
const TIME_THRESHOLD_REGEX = /^(\d+)(y|mo|d|h|min|s|ms)$/;
const parseTimeThreshold = (value) => {
  const matched = value.match(TIME_THRESHOLD_REGEX);
  return matched ? [+matched[1], matched[2]] : void 0;
};
const MULTIPLIERS = {
  y: 31556926e3,
  mo: 26298e5,
  d: 864e5,
  h: 36e5,
  min: 6e4,
  s: 1e3,
  ms: 1
};
const convertTimeThreshold = (timeThresholdExpr) => {
  const parsedValue = parseTimeThreshold(timeThresholdExpr);
  if (!parsedValue) {
    return void 0;
  }
  const [value, unit] = parsedValue;
  return value * MULTIPLIERS[unit];
};
