// @dynatrace/strato-components/charts/core/utils/validators.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var validators_exports = {};
__export(validators_exports, {
  RANGE_MATCHERS: () => RANGE_MATCHERS,
  VALUE_MATCHERS: () => VALUE_MATCHERS,
  buildValidator: () => buildValidator,
  isRangeOperator: () => isRangeOperator
});
module.exports = __toCommonJS(validators_exports);
const isSameType = (a, b) => {
  return typeof a === typeof b;
};
const VALUE_MATCHERS = {
  equals: (subject, value) => `${subject}` === `${value}`,
  "not-equals": (subject, value) => `${subject}` !== `${value}`,
  "less-than": (subject, value) => isSameType(subject, value) && subject < value,
  "less-or-equal": (subject, value) => isSameType(subject, value) && subject <= value,
  "greater-than": (subject, value) => isSameType(subject, value) && subject > value,
  "greater-or-equal": (subject, value) => isSameType(subject, value) && subject >= value,
  "starts-with": (subject, value) => `${subject}`.startsWith(`${value}`),
  "not-starts-with": (subject, value) => !`${subject}`.startsWith(`${value}`),
  "ends-with": (subject, value) => `${subject}`.endsWith(`${value}`),
  "not-ends-with": (subject, value) => !`${subject}`.endsWith(`${value}`),
  contains: (subject, value) => `${subject}`.includes(`${value}`),
  "not-contains": (subject, value) => !`${subject}`.includes(`${value}`),
  "matches-phrase": (subject, value) => subject === value,
  "not-matches-phrase": (subject, value) => subject !== value
};
const RANGE_MATCHERS = {
  "less-than": (subject, value) => isSameType(subject.to, value) && subject.to < value,
  "less-or-equal": (subject, value) => isSameType(subject.to, value) && subject.to <= value,
  "greater-than": (subject, value) => isSameType(subject.to, value) && subject.from > value,
  "greater-or-equal": (subject, value) => isSameType(subject.to, value) && subject.from >= value,
  equals: (subject, value) => subject.to === value && subject.from === value,
  "not-equals": (subject, value) => !(subject.to === value && subject.from === value)
};
const isRangeOperator = (operator) => {
  return operator in RANGE_MATCHERS;
};
function buildValidator(matchers) {
  return (subject, comparator, value) => {
    if (comparator in matchers) {
      return matchers[comparator](
        subject,
        value
      );
    }
    return false;
  };
}
