// @dynatrace/strato-components/charts/core/utils/zoom.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var zoom_exports = {};
__export(zoom_exports, {
  getA11yToolbarConfig: () => getA11yToolbarConfig,
  handlePanningMovement: () => handlePanningMovement,
  isDomainMaxZoom: () => isDomainMaxZoom,
  isDomainMinZoom: () => isDomainMinZoom,
  isSameDomain: () => isSameDomain,
  isZoomBoundaries: () => isZoomBoundaries
});
module.exports = __toCommonJS(zoom_exports);
var import_lodash_es = require("lodash");
var import_constants = require("../constants.js");
const PAN_LARGE_STEP = 0.02;
const PAN_SMALL_STEP = 0.01;
function isDomainMinZoom(newDomain, infiniteZoom, initialDomain) {
  if (newDomain[0] <= +import_constants.GRAIL_MIN_DOMAIN && newDomain[1] >= +import_constants.GRAIL_MAX_DOMAIN) {
    return true;
  }
  if (initialDomain === void 0 || infiniteZoom) {
    return false;
  }
  if (newDomain[0] === initialDomain[0] && newDomain[1] !== initialDomain[1] || newDomain[0] !== initialDomain[0] && newDomain[1] === initialDomain[1]) {
    return false;
  }
  return !((0, import_lodash_es.inRange)(newDomain[0], initialDomain[0], initialDomain[1]) && (0, import_lodash_es.inRange)(newDomain[1], initialDomain[0], initialDomain[1]));
}
function isDomainMaxZoom(newDomain, initialDomain) {
  const newMinValue = newDomain[0];
  const newMaxValue = newDomain[1];
  const isEqualDomain = newMinValue === newMaxValue;
  if (isEqualDomain) {
    return true;
  }
  return initialDomain[0] > initialDomain[1] ? newMaxValue > newMinValue : newMinValue > newMaxValue;
}
function isZoomBoundaries(initialDomain, nextDomain, infiniteZoom) {
  const isMinZoom = isDomainMinZoom(nextDomain, infiniteZoom, initialDomain);
  const isMaxZoom = isDomainMaxZoom(nextDomain, initialDomain);
  return { isMinZoom, isMaxZoom };
}
const handlePanningMovement = (xScale, panDirection, stepType, initialXScaleDomain) => {
  const [, width] = xScale.range();
  const directionFactor = panDirection === "left" ? 1 : -1;
  const panStep = (stepType === "small" ? PAN_SMALL_STEP : PAN_LARGE_STEP) * width;
  return panningMove(xScale, initialXScaleDomain, directionFactor * panStep);
};
const panningMove = (xScale, initialXScaleDomain, delta) => {
  const [rangeMin, rangeMax] = xScale.range();
  const [initialMin, initialMax] = initialXScaleDomain ?? [];
  const newMin = xScale.invert(rangeMin - delta);
  const newMax = xScale.invert(rangeMax - delta);
  if (+newMin > +initialMin && +newMax < +initialMax || !initialXScaleDomain) {
    return [newMin, newMax];
  }
  return [xScale.invert(rangeMin), xScale.invert(rangeMax)];
};
function getA11yToolbarConfig(config) {
  const { zoom, zoomX, download, intent } = config;
  const a11YToolbar = [
    ...[{ explore: 0 }],
    ...zoom?.enabled ? [{ "zoom-in": -1 }] : [],
    ...zoom?.enabled ? [{ "zoom-out": -1 }] : [],
    ...zoom?.enabled || zoomX?.enabled ? [{ reset: -1 }] : [],
    ...download?.enabled ? [{ download: -1 }] : [],
    ...intent ? [{ intent: -1 }] : [],
    ...[{ collapse: -1 }]
  ];
  return a11YToolbar;
}
function isSameDomain(currentDomain, initialDomain) {
  if (currentDomain === void 0) {
    return true;
  }
  return +currentDomain[0] === +initialDomain[0] && +currentDomain[1] === +initialDomain[1];
}
