// @dynatrace/strato-components/esm/tables/DataTable/tanstack-table-core-override/src/utils/getGroupedRowModel.js
import { createRow } from "../core/row";
import { flattenBy, getMemoOptions, memo } from "../utils";
function getGroupedRowModel() {
  return (table) => memo(
    () => [table.getState().grouping, table.getPreGroupedRowModel()],
    (grouping, rowModel) => {
      if (!rowModel.rows.length || !grouping.length) {
        rowModel.rows.forEach((row) => {
          row.depth = 0;
          row.parentId = void 0;
        });
        return rowModel;
      }
      const existingGrouping = grouping.filter(
        (columnId) => table.getColumn(columnId)
      );
      const groupedFlatRows = [];
      const groupedRowsById = {};
      const groupUpRecursively = (rows, depth = 0, parentId) => {
        if (depth >= existingGrouping.length) {
          return rows.map((row) => {
            row.depth = depth;
            groupedFlatRows.push(row);
            groupedRowsById[row.id] = row;
            if (row.subRows) {
              row.subRows = groupUpRecursively(
                row.subRows,
                depth + 1,
                row.id
              );
            }
            return row;
          });
        }
        const columnId = existingGrouping[depth];
        const rowGroupsMap = groupBy(rows, columnId);
        const aggregatedGroupedRows = Array.from(rowGroupsMap.entries()).map(
          ([groupingValue, groupedRows2], index) => {
            let id = `${columnId}:${groupingValue}`;
            id = parentId ? `${parentId}>${id}` : id;
            const subRows = groupUpRecursively(groupedRows2, depth + 1, id);
            subRows.forEach((subRow) => {
              subRow.parentId = id;
            });
            const leafRows = depth ? flattenBy(groupedRows2, (row2) => row2.subRows) : groupedRows2;
            const row = createRow(
              table,
              id,
              leafRows[0].original,
              index,
              depth,
              void 0,
              parentId
            );
            Object.assign(row, {
              groupingColumnId: columnId,
              groupingValue,
              subRows,
              leafRows,
              getValue: (columnId2) => {
                if (existingGrouping.includes(columnId2)) {
                  if (row._valuesCache.hasOwnProperty(columnId2)) {
                    return row._valuesCache[columnId2];
                  }
                  if (groupedRows2[0]) {
                    row._valuesCache[columnId2] = groupedRows2[0].getValue(columnId2) ?? void 0;
                  }
                  return row._valuesCache[columnId2];
                }
                if (row._groupingValuesCache.hasOwnProperty(columnId2)) {
                  return row._groupingValuesCache[columnId2];
                }
                const column = table.getColumn(columnId2);
                const aggregateFn = column?.getAggregationFn();
                if (aggregateFn) {
                  row._groupingValuesCache[columnId2] = aggregateFn(
                    columnId2,
                    leafRows,
                    groupedRows2
                  );
                  return row._groupingValuesCache[columnId2];
                }
              }
            });
            subRows.forEach((subRow) => {
              groupedFlatRows.push(subRow);
              groupedRowsById[subRow.id] = subRow;
            });
            return row;
          }
        );
        return aggregatedGroupedRows;
      };
      const groupedRows = groupUpRecursively(rowModel.rows, 0);
      groupedRows.forEach((subRow) => {
        groupedFlatRows.push(subRow);
        groupedRowsById[subRow.id] = subRow;
      });
      return {
        rows: groupedRows,
        flatRows: groupedFlatRows,
        rowsById: groupedRowsById
      };
    },
    getMemoOptions(table.options, "debugTable", "getGroupedRowModel", () => {
      table._queue(() => {
        table._autoResetExpanded();
        table._autoResetPageIndex();
      });
    })
  );
}
function groupBy(rows, columnId) {
  const groupMap = /* @__PURE__ */ new Map();
  return rows.reduce((map, row) => {
    const resKey = `${row.getGroupingValue(columnId)}`;
    const previous = map.get(resKey);
    if (!previous) {
      map.set(resKey, [row]);
    } else {
      previous.push(row);
    }
    return map;
  }, groupMap);
}
export {
  getGroupedRowModel
};
