// @dynatrace/strato-components/esm/tables/DataTable/tanstack-table-core-override/src/utils/getSortedRowModel.js
import { getMemoOptions, memo } from "../utils";
function getSortedRowModel() {
  return (table) => memo(
    () => [table.getState().sorting, table.getPreSortedRowModel()],
    (sorting, rowModel) => {
      if (!rowModel.rows.length || !sorting?.length) {
        return rowModel;
      }
      const sortingState = table.getState().sorting;
      const sortedFlatRows = [];
      const availableSorting = sortingState.filter(
        (sort) => table.getColumn(sort.id)?.getCanSort()
      );
      const columnInfoById = {};
      availableSorting.forEach((sortEntry) => {
        const column = table.getColumn(sortEntry.id);
        if (!column) return;
        columnInfoById[sortEntry.id] = {
          sortUndefined: column.columnDef.sortUndefined,
          invertSorting: column.columnDef.invertSorting,
          sortingFn: column.getSortingFn()
        };
      });
      const sortData = (rows) => {
        const sortedData = rows.map((row) => row.clone());
        sortedData.sort((rowA, rowB) => {
          for (let i = 0; i < availableSorting.length; i += 1) {
            const sortEntry = availableSorting[i];
            const columnInfo = columnInfoById[sortEntry.id];
            const sortUndefined = columnInfo.sortUndefined;
            const isDesc = sortEntry?.desc ?? false;
            let sortInt = 0;
            if (sortUndefined) {
              const aValue = rowA.getValue(sortEntry.id);
              const bValue = rowB.getValue(sortEntry.id);
              const aUndefined = aValue === void 0;
              const bUndefined = bValue === void 0;
              if (aUndefined || bUndefined) {
                if (sortUndefined === "first") return aUndefined ? -1 : 1;
                if (sortUndefined === "last") return aUndefined ? 1 : -1;
                sortInt = aUndefined && bUndefined ? 0 : aUndefined ? sortUndefined : -sortUndefined;
              }
            }
            if (sortInt === 0) {
              sortInt = columnInfo.sortingFn(rowA, rowB, sortEntry.id);
            }
            if (sortInt !== 0) {
              if (isDesc) {
                sortInt *= -1;
              }
              if (columnInfo.invertSorting) {
                sortInt *= -1;
              }
              return sortInt;
            }
          }
          return rowA.index - rowB.index;
        });
        sortedData.forEach((row) => {
          sortedFlatRows.push(row);
          if (row.subRows?.length) {
            row.subRows = sortData(row.subRows);
          }
        });
        return sortedData;
      };
      return {
        rows: sortData(rowModel.rows),
        flatRows: sortedFlatRows,
        rowsById: rowModel.rowsById
      };
    },
    getMemoOptions(
      table.options,
      "debugTable",
      "getSortedRowModel",
      () => table._autoResetPageIndex()
    )
  );
}
export {
  getSortedRowModel
};
