// @dynatrace/strato-components/esm/charts/core/components/interactions-overlay/interactions/explore/dragInteractions.js
import { SELECTION_RIGHT_OVERFLOW } from "../../constants.js";
import { capSelectionAtBoundaries } from "../../utils/cap-selection-at-boundaries.js";
import { centerPositionVertically } from "../../utils/center-position-vertically.js";
import { normalizeSelectionArea } from "../../utils/normalize-selection-area.js";
const exploreDragStart = ({
  event,
  onSelectStart
}) => {
  onSelectStart?.({
    triggeredBy: "pointer",
    position: event.startHoldingAt
  });
};
const exploreDragMove = ({
  event,
  height,
  width,
  onSelectChange
}) => {
  const {
    startHoldingAt: startHoldingAtProp,
    currentPosition: currentPositionProp
  } = event;
  const startHoldingAt = centerPositionVertically(startHoldingAtProp, height);
  const currentPosition = centerPositionVertically(currentPositionProp, height);
  const [start, end] = normalizeSelectionArea([
    startHoldingAt,
    currentPosition
  ]);
  onSelectChange?.({
    triggeredBy: "pointer",
    start,
    end,
    mode: "explore"
  });
};
const exploreDragEnd = ({
  event,
  height,
  width,
  onSelectEnd,
  lastMousePositionRef
}) => {
  const { startHoldingAt: startHoldingAtProp, releasedAt: releasedAtProp } = event;
  const startHoldingAt = centerPositionVertically(startHoldingAtProp, height);
  const releasedAt = centerPositionVertically(releasedAtProp, height);
  const normalizedSelectionArea = normalizeSelectionArea([
    startHoldingAt,
    releasedAt
  ]);
  const { start, end } = capSelectionAtBoundaries(
    normalizedSelectionArea,
    width + SELECTION_RIGHT_OVERFLOW
  );
  if (start.relative.x < 0 || end.relative.x > width + SELECTION_RIGHT_OVERFLOW) {
    return;
  }
  onSelectEnd?.({
    triggeredBy: "pointer",
    start,
    end
  });
  lastMousePositionRef.current = releasedAt;
};
export {
  exploreDragEnd,
  exploreDragMove,
  exploreDragStart
};
