// @dynatrace/strato-components/esm/charts/core/components/interactions-overlay/interactions/explore/keyboardInteractions.js
import { isUndefined } from "lodash-es";
const exploreKeyboardEscape = ({
  selectionArea,
  eventCancelled,
  onSelectEnd,
  lastMousePositionRef,
  pinned,
  onHover,
  onTogglePin,
  onModeChange
}) => {
  if (!isUndefined(selectionArea)) {
    eventCancelled.current = true;
    onSelectEnd?.({ cancelled: true });
    if (lastMousePositionRef.current) {
      onHover?.({
        triggeredBy: "pointer",
        position: lastMousePositionRef.current
      });
    }
  }
  if (pinned) {
    onTogglePin?.({
      triggeredBy: "keyboard",
      pin: false,
      position: lastMousePositionRef.current
    });
  }
  onModeChange?.({ mode: "explore", keyboard: false });
};
const exploreKeyboardEnter = ({
  selectionArea,
  width,
  onSelectEnd,
  onZoomChange,
  isZoomXDisabled
}) => {
  if (!isZoomXDisabled && selectionArea) {
    const deltaStart = selectionArea[0].relative.x;
    const deltaEnd = selectionArea[1].relative.x - width;
    onSelectEnd?.({ cancelled: true });
    onZoomChange?.({
      triggeredBy: "keyboard",
      deltaStart,
      deltaEnd,
      behavior: "zoom-x"
    });
  }
};
export {
  exploreKeyboardEnter,
  exploreKeyboardEscape
};
