// @dynatrace/strato-components/esm/charts/timeseries/providers/TimeseriesFilteredDataProvider.js
import { jsx } from "react/jsx-runtime";
import { safeGet } from "../../core/utils/safe-get.js";
import { TimeseriesFilteredDataContext } from "../context/TimeseriesFilteredData.context.js";
import { useFilteredSeriesIdsState } from "../hooks/store/selectors.js";
import { useBuildFilteredMetadata } from "../hooks/useBuildFilteredMetadata.js";
import { useTimeseriesData } from "../hooks/useTimeseriesData.js";
import {
  getDatapointTime,
  mergeTimeseriesFromVariants
} from "../utils/timeseries.js";
function binarySearchOrder(a, b) {
  return getDatapointTime(a.start, a.end) - getDatapointTime(b.start, b.end);
}
const TimeseriesFilteredDataProvider = ({ children }) => {
  const {
    dataPerVariant,
    seriesMap: initialSeriesMap,
    valueRepPerVariant,
    units
  } = useTimeseriesData();
  const filteredSeriesIds = useFilteredSeriesIdsState();
  const hasId = (series) => filteredSeriesIds.has(safeGet(initialSeriesMap, series).id);
  const filteredTimeseriesPerVariant = filteredSeriesIds.size > 0 ? {
    ...dataPerVariant,
    line: dataPerVariant["line"].filter(hasId),
    bar: dataPerVariant["bar"].filter(hasId),
    area: dataPerVariant["area"].filter(hasId),
    band: dataPerVariant["band"].filter(hasId)
  } : dataPerVariant;
  const {
    seriesMap,
    datapointsMap,
    filteredUnits,
    filteredTimeseries,
    originalFilteredUnits
  } = useBuildFilteredMetadata(
    filteredTimeseriesPerVariant,
    valueRepPerVariant,
    initialSeriesMap,
    units
  );
  const data = mergeTimeseriesFromVariants(filteredTimeseries);
  const allDatapoints = data.flatMap((series) => series.datapoints).sort(binarySearchOrder);
  return /* @__PURE__ */ jsx(
    TimeseriesFilteredDataContext.Provider,
    {
      value: {
        data,
        dataPerVariant: filteredTimeseries,
        seriesMap,
        datapointsMap,
        units: filteredUnits,
        originalUnits: originalFilteredUnits,
        allDatapoints
      },
      children
    }
  );
};
export {
  TimeseriesFilteredDataProvider
};
