// @dynatrace/strato-components/esm/charts/timeseries/providers/TimeseriesDataProvider.js
import { jsx } from "react/jsx-runtime";
import { uniq } from "lodash-es";
import { useMemo } from "react";
import { useColorRule } from "../../core/color-rule/useColorRule.js";
import { getColorPalette } from "../../core/utils/get-color-palette.js";
import { TimeseriesDataContext } from "../context/TimeseriesData.context.js";
import { useColorPalette } from "../hooks/useColorPalette.js";
import { useDataByYAxis } from "../hooks/useDataByYAxis.js";
import { useGapPolicy } from "../hooks/useGapPolicy.js";
import { usePointsDisplay } from "../hooks/usePointsDisplay.js";
import { mergeTimeseriesFromVariants } from "../utils/timeseries.js";
function getAllUnits(data) {
  return uniq(data.shapePriority.flatMap((v) => data[v]?.map((s) => s.unit)));
}
function useBuildInitialMetadata(data, valueRepresentation, mergedData) {
  const colorPalette = useColorPalette();
  const gapPolicy = useGapPolicy();
  const pointsDisplay = usePointsDisplay();
  const seriesMap = /* @__PURE__ */ new WeakMap();
  const allUnits = getAllUnits(data);
  const colorRules = useColorRule(mergedData, "timeseries");
  let startingIndex = 0;
  const valueRepPerVariant = /* @__PURE__ */ new Map();
  for (const variant of data.shapePriority) {
    const variantSeries = data[variant];
    for (let i = 0; i < variantSeries.length; i++) {
      const series = variantSeries[i];
      const id = crypto.randomUUID();
      const color = series.color ? { Default: series.color } : colorRules.for(series) ?? getColorPalette(
        Array.isArray(series.name) ? series.name[0] : series.name,
        startingIndex + i,
        colorPalette
      );
      const metadata = {
        id,
        color,
        gapPolicy: series.gapPolicy ?? gapPolicy,
        pointsDisplay: series.pointsDisplay ?? pointsDisplay
      };
      if (!valueRepPerVariant.get(variant) && series.valueRepresentation) {
        valueRepPerVariant.set(variant, series.valueRepresentation);
      }
      seriesMap.set(series, metadata);
    }
    if (!valueRepPerVariant.get(variant)) {
      valueRepPerVariant.set(variant, valueRepresentation);
    }
    startingIndex += variantSeries.length;
  }
  return {
    units: allUnits,
    seriesMap,
    valueRepPerVariant
  };
}
function cloneSeries(array) {
  return array.map((series) => ({
    ...series,
    datapoints: series.datapoints.map((datapoint) => ({ ...datapoint }))
  }));
}
const TimeseriesDataProvider = ({
  children,
  data: _data,
  valueRepresentation,
  onMessage
}) => {
  const clonedData = {
    line: cloneSeries(_data.line),
    bar: cloneSeries(_data.bar),
    area: cloneSeries(_data.area),
    band: cloneSeries(_data.band),
    shapePriority: _data.shapePriority
  };
  const data = useDataByYAxis(clonedData);
  const mergedData = useMemo(() => mergeTimeseriesFromVariants(data), [data]);
  const { units, seriesMap, valueRepPerVariant } = useBuildInitialMetadata(
    data,
    valueRepresentation,
    mergedData
  );
  return /* @__PURE__ */ jsx(
    TimeseriesDataContext.Provider,
    {
      value: {
        data: mergedData,
        dataPerVariant: data,
        valueRepPerVariant,
        seriesMap,
        units
      },
      children
    }
  );
};
TimeseriesDataProvider["displayName"] = "TimeseriesDataProvider";
export {
  TimeseriesDataProvider,
  useBuildInitialMetadata
};
