// @dynatrace/strato-components/esm/editors/base-code-editor/BaseCodeEditor.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { closeCompletion } from "@codemirror/autocomplete";
import {
  foldedRanges,
  forceParsing,
  indentUnit,
  syntaxTreeAvailable
} from "@codemirror/language";
import {
  Compartment,
  EditorSelection,
  EditorState,
  Prec
} from "@codemirror/state";
import { EditorView } from "@codemirror/view";
import clsx from "clsx";
import {
  forwardRef,
  useCallback,
  useEffect,
  useImperativeHandle,
  useLayoutEffect,
  useRef,
  useState
} from "react";
import { useIntl } from "react-intl";
import * as styles from "./BaseCodeEditor.sty.js";
import { configureLineWrap } from "./configuration/configure-line-wrap.js";
import { configurePlaceholder } from "./configuration/configure-placeholder.js";
import { configureReadOnly } from "./configuration/configure-read-only.js";
import { configureSpellCheck } from "./configuration/configure-spell-check.js";
import {
  getMultipleLinesDefaultConfiguration,
  getSingleLineDefaultConfiguration,
  getDefaultConfiguration
} from "./configuration/getDefaultConfigurations.js";
import { getLanguageConfiguration } from "./configuration/language-extensions.js";
import { highlightExtension } from "./highlight/highlight.extension.js";
import {
  dataTestId as dataTestIdAttr,
  baseCodeEditorTestId,
  baseCodeEditorRootTestId,
  baseCodeEditorScrollerTestId
} from "./testIds.js";
import { calculateFolding } from "./utils/calculate-folding.js";
import { getKeyMaps } from "./utils/key-maps.js";
import { runFoldings } from "./utils/run-foldings.js";
import { sanitizeText } from "./utils/sanitize-text.js";
import { sanitizeValue } from "./utils/sanitize-value.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useControllableState } from "../../core/hooks/useControllableState.js";
import { useFormControlState } from "../../core/hooks/useFormControlState.js";
import { useMergeRefs } from "../../core/hooks/useMergeRefs.js";
import { useFocusRing } from "../../core/styles/useFocusRing.js";
import { InternalBaseItem as BaseItem } from "../../forms/forms-core/components/form-field-messages/FormFieldMessageBaseItem.js";
import { useSyncedFormControlProps } from "../../forms/forms-core/hooks/useSyncedFormControlProps.js";
import { inputGroupClassName } from "../../layouts/input-group/InputGroup.sty.js";
EditorView.EDIT_CONTEXT = false;
const languageCompartment = new Compartment();
const editorBaseConfigCompartment = new Compartment();
const lineWrappingCompartment = new Compartment();
const keymapCompartment = new Compartment();
const placeholderCompartment = new Compartment();
const highlightOnChangeCompartment = new Compartment();
const getEditorConfiguration = (intlProps, options) => {
  const { singleLine, gutterConfiguration, hideLineNumbers, required, size } = options;
  const isRequired = required ?? false;
  const editorSize = size ?? "default";
  if (singleLine) {
    return getSingleLineDefaultConfiguration(intlProps, {
      required: isRequired,
      size: editorSize
    });
  }
  if (hideLineNumbers) {
    return getDefaultConfiguration(intlProps, {
      required: isRequired,
      size: editorSize
    });
  }
  return getMultipleLinesDefaultConfiguration(intlProps, {
    gutterConfiguration,
    required: isRequired,
    size: editorSize
  });
};
function useLazyEditorViewRef() {
  const viewRef = useRef(null);
  if (viewRef.current === null) {
    viewRef.current = new EditorView({
      state: EditorState.create()
    });
  }
  return viewRef;
}
const getDomEventHandlers = (onFocus, onBlur) => {
  return EditorView.domEventHandlers({
    // Prevent the Escape key from propagating beyond the editor.
    // This is necessary for editors that are placed inside an overlay (e.g. Modal),
    // where pressing `Escape` to exit the edit mode would otherwise close the overlay.
    keydown: (event) => {
      if (event.key === "Escape") {
        event.stopPropagation();
      }
    },
    // when the editor content element receives focus, call the provided focus handler
    focus: (event) => {
      onFocus?.(event);
    },
    // when the editor content element loses focus, call the provided blur handler
    blur: (event) => {
      onBlur?.(event);
    }
  });
};
const BaseCodeEditor = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const {
    controlState,
    value,
    readOnly = false,
    spellCheck = false,
    lineWrap = false,
    onChange,
    onFoldingChange,
    defaultFolding,
    folding,
    fullHeight = false,
    placeholder,
    configuration,
    singleLine = false,
    hideLineNumbers = false,
    size = "default",
    onFocus,
    onBlur,
    gutterConfiguration,
    id: propId,
    required = false,
    style: consumerStyle,
    className: consumerClassName,
    "aria-label": ariaLabel,
    "aria-labelledby": ariaLabelledBy,
    "aria-details": ariaDetails,
    "aria-describedby": ariaDescribedBy,
    "data-testid": dataTestId = baseCodeEditorRootTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    // @ts-expect-error - We want to allow overriding data-dt-component as we use the component internally in other Strato components, but it should not appear in the public API.
    "data-dt-component": dataDtComponent,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
  const intl = useIntl();
  const [hasError, hint] = useFormControlState(controlState);
  const containerRef = useRef(null);
  const rootRef = useMergeRefs([containerRef]);
  const sanitizedOnChange = useCallback(
    (value2) => onChange?.(sanitizeText(value2)),
    [onChange]
  );
  const viewRef = useLazyEditorViewRef();
  const { id } = useSyncedFormControlProps({
    propId
  });
  const valueChangeRef = useRef(void 0);
  const valueRef = useRef(value ?? "");
  const isMarkedDirtyRef = useRef(false);
  const prevEditorConfigRef = useRef(void 0);
  const prevLangConfigRef = useRef(void 0);
  const [previousValue, setPreviousValue] = useState("");
  const transactionRef = useRef(void 0);
  const { focusProps, isFocusVisible } = useFocusRing({
    color: hasError ? "critical" : "neutral"
  });
  const [controlledFolding] = useControllableState({
    defaultValue: defaultFolding,
    onChange: (values) => {
      if (values) {
        onFoldingChange?.(values);
      }
    },
    value: folding
  });
  useImperativeHandle(
    forwardedRef,
    () => ({
      element: containerRef.current,
      editorView: () => viewRef.current,
      focus: () => viewRef.current.focus(),
      selectRange: (range) => {
        const editorView = viewRef.current;
        if (editorView && range) {
          const docLength = editorView.state.doc.length;
          const from = range.from < docLength ? range.from : 0;
          const to = range.to < docLength ? range.to : docLength;
          const transaction = editorView.state.update({
            selection: EditorSelection.range(to, from),
            scrollIntoView: true
          });
          editorView.dispatch(transaction);
          editorView.focus();
        }
      },
      setCursorPosition: (position) => {
        viewRef.current.dispatch({
          selection: EditorSelection.cursor(
            typeof position === "number" ? Math.min(
              Math.max(position, 0),
              viewRef.current.state.doc.length
            ) : viewRef.current.state.doc.length
          ),
          scrollIntoView: true
        });
        viewRef.current.focus();
      }
    }),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    []
  );
  useEffect(() => {
    const editorView = viewRef.current;
    containerRef.current?.appendChild(editorView.dom);
    editorView.dom.setAttribute(dataTestIdAttr, baseCodeEditorTestId);
    return () => editorView.destroy();
  }, []);
  useEffect(() => {
    const { contentDOM } = viewRef.current;
    const ariaLabelAttribute = "aria-label";
    const ariaLabelledByAttribute = "aria-labelledby";
    const ariaDetailsAttribute = "aria-details";
    const ariaDescribedByAttribute = "aria-describedby";
    if (ariaDescribedBy) {
      contentDOM.setAttribute(ariaDescribedByAttribute, ariaDescribedBy);
    }
    if (ariaDetails) {
      contentDOM.setAttribute(ariaDetailsAttribute, ariaDetails);
    }
    if (ariaLabelledBy) {
      contentDOM.setAttribute(ariaLabelledByAttribute, ariaLabelledBy);
    } else {
      contentDOM.setAttribute(
        ariaLabelAttribute,
        ariaLabel ?? intl.formatMessage({
          defaultMessage: "Code Editor",
          id: "EBcYgbL0S1T6EpS2",
          description: "Aria-label for the code editors."
        })
      );
    }
    return () => [
      ariaLabelAttribute,
      ariaLabelledByAttribute,
      ariaDetailsAttribute,
      ariaDescribedByAttribute
    ].forEach((ariaAttribute) => contentDOM.removeAttribute(ariaAttribute));
  }, [ariaLabel, ariaLabelledBy, ariaDetails, ariaDescribedBy, intl]);
  useLayoutEffect(() => {
    viewRef.current.requestMeasure();
  }, [viewRef, size]);
  const { messages, defaultLocale, locale } = useIntl();
  useEffect(() => {
    const editorView = viewRef.current;
    const handleValueChange = (transaction) => {
      const newValue = transaction.newDoc.toString();
      if (valueRef.current !== newValue) {
        transactionRef.current = transaction;
        valueRef.current = newValue;
        if (isMarkedDirtyRef.current) {
          isMarkedDirtyRef.current = false;
        } else {
          valueChangeRef.current?.(newValue);
        }
      } else {
        isMarkedDirtyRef.current = false;
      }
      return transaction;
    };
    editorView.setState(
      EditorState.create({
        // Initial value of the editor
        doc: value ?? "",
        extensions: [
          getDomEventHandlers(onFocus, onBlur),
          Prec.highest(keymapCompartment.of(getKeyMaps(viewRef))),
          languageCompartment.of([]),
          editorBaseConfigCompartment.of([]),
          // Indent unit is two spaces instead of tab
          indentUnit.of("  "),
          // Filter for transactions that gets applied before they are processed. This allows us
          // to control which transactions go through as well as adding/editing transactions if we please.
          EditorState.transactionFilter.of((transaction) => {
            if (transaction.docChanged) {
              return handleValueChange(transaction);
            }
            return transaction;
          }),
          EditorView.updateListener.of((update) => {
            const isUndoOrRedo = update.transactions.some(
              (transaction) => transaction.isUserEvent("undo") || transaction.isUserEvent("redo")
            );
            if (update.geometryChanged && !update.docChanged && !update.focusChanged && update.heightChanged) {
              const ranges = foldedRanges(update.state);
              const foldValues = calculateFolding(ranges);
              onFoldingChange?.(foldValues);
            } else if (update.docChanged && isUndoOrRedo) {
              update.transactions.forEach(
                (transaction) => handleValueChange(transaction)
              );
            }
          }),
          lineWrappingCompartment.of([]),
          placeholderCompartment.of([]),
          highlightOnChangeCompartment.of([])
        ]
      })
    );
  }, []);
  useEffect(() => {
    const editorView = viewRef.current;
    const prev = prevEditorConfigRef.current;
    if (prev !== void 0 && prev.singleLine === singleLine && prev.hideLineNumbers === hideLineNumbers && prev.gutterConfiguration === gutterConfiguration && prev.required === required && prev.size === size && prev.locale === locale && prev.defaultLocale === defaultLocale && prev.messages === messages) {
      return;
    }
    prevEditorConfigRef.current = {
      singleLine,
      hideLineNumbers,
      gutterConfiguration,
      required,
      size,
      locale,
      defaultLocale,
      messages
    };
    const intlProps = { locale, defaultLocale, messages };
    editorView.dispatch({
      effects: editorBaseConfigCompartment.reconfigure(
        getEditorConfiguration(intlProps, {
          singleLine,
          hideLineNumbers,
          gutterConfiguration,
          required,
          size
        })
      )
    });
  }, [
    singleLine,
    hideLineNumbers,
    gutterConfiguration,
    required,
    size,
    locale,
    defaultLocale,
    messages
  ]);
  useEffect(() => {
    const editorView = viewRef.current;
    const prev = prevLangConfigRef.current;
    if (prev !== void 0 && prev.language === configuration?.language && prev.extendWith === configuration?.extendWith && prev.autocompleteOptions === configuration?.autocompleteOptions && prev.locale === locale && prev.defaultLocale === defaultLocale && prev.messages === messages) {
      return;
    }
    prevLangConfigRef.current = {
      language: configuration?.language,
      extendWith: configuration?.extendWith,
      autocompleteOptions: configuration?.autocompleteOptions,
      locale,
      defaultLocale,
      messages
    };
    const intlProps = { locale, defaultLocale, messages };
    editorView.dispatch({
      effects: languageCompartment.reconfigure(
        getLanguageConfiguration(configuration || {}, intlProps)
      )
    });
  }, [configuration, locale, defaultLocale, messages]);
  useEffect(() => {
    configureReadOnly(readOnly, viewRef, keymapCompartment);
  }, [readOnly]);
  useEffect(() => {
    configureLineWrap(lineWrap, viewRef, lineWrappingCompartment);
  }, [lineWrap]);
  useEffect(() => {
    configurePlaceholder(placeholder, viewRef, placeholderCompartment);
  }, [placeholder]);
  useEffect(() => {
    configureSpellCheck(spellCheck, readOnly, viewRef);
  }, [viewRef, readOnly, spellCheck]);
  useEffect(() => {
    valueChangeRef.current = onChange !== void 0 ? sanitizedOnChange : void 0;
  }, [onChange, sanitizedOnChange]);
  useEffect(() => {
    const sanitizedValue = sanitizeValue(value);
    if (viewRef.current.state.doc.toString() === sanitizedValue) {
      return;
    }
    setPreviousValue(viewRef.current.state.doc.toString());
    isMarkedDirtyRef.current = true;
    try {
      let selection;
      if (!viewRef.current.hasFocus) {
        selection = void 0;
      } else {
        const editorLineCount = viewRef.current.state.doc.lines;
        const lastLine = viewRef.current.state.doc.line(editorLineCount);
        const selectionPosition = Math.max(0, lastLine.length - 1);
        selection = { anchor: selectionPosition, head: selectionPosition };
      }
      transactionRef.current = void 0;
      const transaction = viewRef.current.state.update({
        changes: [
          {
            from: 0,
            to: viewRef.current.state.doc.length,
            insert: sanitizedValue
          }
        ],
        selection
      });
      viewRef.current.dispatch(transaction);
    } catch {
    }
  }, [value]);
  useEffect(() => {
    const editorView = viewRef.current;
    editorView.dispatch({
      effects: highlightOnChangeCompartment.reconfigure([
        highlightExtension(previousValue)
      ])
    });
  }, [previousValue]);
  useEffect(() => {
    runFoldings(controlledFolding ?? [], viewRef.current);
  }, [controlledFolding]);
  useEffect(() => {
    const codeEditorInput = viewRef.current.dom.querySelector(
      "div[contenteditable]"
    );
    if (codeEditorInput) {
      hasError ? codeEditorInput.setAttribute("aria-invalid", "true") : codeEditorInput.removeAttribute("aria-invalid");
      codeEditorInput.tabIndex = -1;
    }
  }, [hasError]);
  const handleOutsideEditorClick = useCallback(
    (e) => {
      const autoCompleteContainer = document.body.querySelector(
        ".cm-tooltip-autocomplete"
      );
      if (!autoCompleteContainer?.contains(e.target)) {
        closeCompletion(viewRef.current);
      }
    },
    [viewRef]
  );
  useEffect(() => {
    document.addEventListener("pointerdown", handleOutsideEditorClick);
    return () => {
      document.removeEventListener("pointerdown", handleOutsideEditorClick);
    };
  }, [handleOutsideEditorClick]);
  useEffect(() => {
    if (!value) {
      return;
    }
    let timeoutId = null;
    let done = false;
    const cleanup = () => {
      if (timeoutId) {
        window.clearTimeout(timeoutId);
      }
    };
    const tick = () => {
      const wholeSyntaxTreeParsed = syntaxTreeAvailable(
        viewRef.current.state,
        viewRef.current.state.doc.length
      );
      if (wholeSyntaxTreeParsed) {
        cleanup();
        return;
      }
      done = forceParsing(
        viewRef.current,
        viewRef.current.state.doc.length,
        100
      );
      if (!done) {
        timeoutId = window.setTimeout(tick, 10);
      }
    };
    tick();
    return cleanup;
  }, [value]);
  useEffect(() => {
    const editorScrollContainer = viewRef.current.dom.querySelector(".cm-scroller");
    editorScrollContainer?.setAttribute("tabIndex", "0");
    editorScrollContainer?.setAttribute("role", "group");
    editorScrollContainer?.setAttribute(
      dataTestIdAttr,
      baseCodeEditorScrollerTestId
    );
    const handleFocus = (event) => {
      if (event instanceof FocusEvent) {
        focusProps.onFocus(event);
      }
    };
    const handleBlur = (event) => {
      if (event instanceof FocusEvent) {
        focusProps.onBlur(event);
      }
    };
    const handleKeyDown = (event) => {
      if (event instanceof KeyboardEvent) {
        if (event.key === "Enter" && !event.metaKey && !event.ctrlKey && !readOnly) {
          event.preventDefault();
          viewRef.current.focus();
        }
      }
    };
    editorScrollContainer?.addEventListener("focus", handleFocus);
    editorScrollContainer?.addEventListener("blur", handleBlur);
    editorScrollContainer?.addEventListener("keydown", handleKeyDown);
    return () => {
      editorScrollContainer?.removeEventListener("focus", handleFocus);
      editorScrollContainer?.removeEventListener("blur", handleBlur);
      editorScrollContainer?.removeEventListener("keydown", handleKeyDown);
    };
  }, [focusProps, readOnly, viewRef]);
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(
      "div",
      {
        ref: rootRef,
        className: clsx(
          styles.editorContainer({
            focusVisible: isFocusVisible,
            fullHeight,
            hasError,
            lineNumbers: singleLine || hideLineNumbers ? "hidden" : void 0,
            readOnly,
            size
          }),
          inputGroupClassName,
          consumerClassName
        ),
        style: consumerStyle,
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        ...behavioralTrackingProps,
        ...restProps,
        "data-dt-component": dataDtComponent
      }
    ),
    /* @__PURE__ */ jsx(
      "input",
      {
        type: "text",
        id,
        disabled: readOnly,
        tabIndex: -1,
        onFocus: () => (
          // forward focus to the editor
          // we need the hidden input unfortunately since the editor does not support focus properly
          viewRef.current.focus()
        ),
        "aria-hidden": "true",
        style: {
          position: "absolute",
          height: 0,
          width: 0,
          opacity: 0,
          margin: 0,
          padding: 0,
          border: 0
        }
      }
    ),
    (hasError || hint) && /* @__PURE__ */ jsx(
      "span",
      {
        "aria-live": "polite",
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        children: /* @__PURE__ */ jsx(BaseItem, { variant: hasError ? "error" : "hint", children: hint })
      }
    )
  ] });
});
BaseCodeEditor.displayName = "BaseCodeEditor";
export {
  BaseCodeEditor
};
