// @dynatrace/strato-components/esm/filters/segment-selector/parts/ShowMorePanel/ShowMorePanel.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import {
  useMemo,
  useState,
  Children
} from "react";
import { FormattedMessage } from "react-intl";
import { isElement } from "react-is";
import {
  ChevronDownSmallIcon,
  ChevronUpSmallIcon
} from "@dynatrace/strato-icons";
import * as styles from "./ShowMorePanel.sty.js";
import { Button } from "../../../../buttons/button/Button.js";
import { isDev } from "../../../../core/utils/is-node-environment.js";
import { Flex } from "../../../../layouts/flex/Flex.js";
const ExtendedContent = ({ children }) => children;
const PrimaryContent = ({ children }) => children;
function useChildrenSlots(children) {
  const memoizedSlots = useMemo(() => {
    const slots = {};
    Children.forEach(children, (child) => {
      if (isElement(child)) {
        switch (child.type) {
          case ExtendedContent:
            slots.extendedContent = child;
            break;
          case PrimaryContent:
          default:
            slots.primaryContent = child;
            break;
        }
      }
    });
    return slots;
  }, [children]);
  return memoizedSlots;
}
const ShowMorePanelRoot = ({ children }) => {
  const { primaryContent, extendedContent } = useChildrenSlots(children);
  const [isExpanded, setIsExpanded] = useState(false);
  if (!children) {
    if (isDev()) {
      console.warn(
        "SegmentSelector: The used child is not supported by the ExpandableSection."
      );
    }
    return null;
  }
  if (!extendedContent) {
    return primaryContent;
  }
  const icon = isExpanded ? /* @__PURE__ */ jsx(ChevronUpSmallIcon, {}) : /* @__PURE__ */ jsx(ChevronDownSmallIcon, {});
  const dataDtProperties = `state:${isExpanded ? "less" : "more"}`;
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx("ul", { className: styles.list, children: primaryContent }),
    isExpanded && /* @__PURE__ */ jsx(
      "ul",
      {
        className: clsx(styles.expandableWrapper, {
          [styles.expandableWrapperExpanded]: isExpanded,
          [styles.expandableWrapperCollapsed]: !isExpanded
        }),
        children: extendedContent
      }
    ),
    /* @__PURE__ */ jsx(
      Button,
      {
        onClick: () => setIsExpanded((previousState) => !previousState),
        width: "full",
        textAlign: "start",
        style: {
          transitionTimingFunction: "none",
          transitionDuration: "0s",
          transition: "none"
        },
        "data-dt-name": "ShowMoreLess",
        "data-dt-properties": dataDtProperties,
        children: /* @__PURE__ */ jsxs(Flex, { gap: 2, as: "span", paddingLeft: 4, children: [
          /* @__PURE__ */ jsx(
            FormattedMessage,
            {
              defaultMessage: "Show {isExpanded, select, true {less} false {more} other{}}",
              id: "mYmd/xMZulbADHsY",
              description: "Text shown on the expanding button within the SegmentSelector preset list, when currently expanded.",
              values: { isExpanded }
            }
          ),
          icon
        ] })
      }
    )
  ] });
};
const ShowMorePanel = Object.assign(ShowMorePanelRoot, {
  PrimaryContent,
  ExtendedContent
});
export {
  ShowMorePanel
};
