// @dynatrace/strato-components/esm/filters/filter-field/utils/conversion/string-tree-conversion.js
import { convertLezerTree } from "./tree-conversion.js";
import { getNumberParser } from "../../parser/getNumberParser.js";
import { parser } from "../../parser/parser.js";
import { toInternalValidatorMap } from "../../utils/validation-utils/normalize-validator-map.js";
function convertStringToFilterFieldTree(filterString, optionsOrValidatorMap, parserConfig) {
  let validatorMap;
  let customTypes = void 0;
  let _parserConfig;
  if (isConvertStringToFilterFieldTreeOptions(optionsOrValidatorMap)) {
    validatorMap = optionsOrValidatorMap?.validatorMap;
    customTypes = optionsOrValidatorMap?.customTypes;
    _parserConfig = optionsOrValidatorMap?.parserConfig;
  } else {
    validatorMap = optionsOrValidatorMap;
    _parserConfig = parserConfig;
  }
  const numberParser = getNumberParser();
  const parsedTree = parser.parse(filterString);
  const { tree, isValid } = convertLezerTree({
    tree: parsedTree,
    numberParser,
    doc: filterString,
    internalValidatorMap: toInternalValidatorMap(validatorMap),
    customTypes,
    parserConfig: _parserConfig
  });
  return { tree, isValid };
}
function isConvertStringToFilterFieldTreeOptions(value) {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    return false;
  }
  return "validatorMap" in value || "customTypes" in value || "parserConfig" in value;
}
export {
  convertStringToFilterFieldTree
};
