// @dynatrace/strato-components/esm/filters/filter-field/utils/conversion/tree-conversion.js
import { sortFilterFieldTree } from "./tree-precedence.js";
import { usedErrors } from "../../editor-state-configs/editor-config.js";
import { convertFilterGroup } from "./lezer-tree-conversion/parent-node/group.js";
const parserVersion = "004";
function convertLezerTree({
  tree,
  view,
  numberParser,
  doc,
  internalValidatorMap,
  customTypes,
  parserConfig
}) {
  const errorsStateField = view?.state.field(usedErrors, false);
  if (!doc || doc === "") {
    errorsStateField && (errorsStateField.errors = []);
    return { isValid: true, isValidSearch: true };
  }
  const { tree: convertedFlatTree, errors: initialConversionErrors } = convertFlatLezerTree({
    tree,
    numberParser,
    doc,
    internalValidatorMap,
    customTypes,
    parserConfig
  });
  const {
    tree: convertedTree,
    errors: sortedTreeErrors,
    isValidSearch
  } = sortFilterFieldTree(convertedFlatTree, parserConfig, doc);
  const errors = [...initialConversionErrors, ...sortedTreeErrors];
  if (errorsStateField) {
    errors.sort((a, b) => a.range.from < b.range.from ? -1 : 1);
    errorsStateField.errors = errors;
  }
  if (convertedTree) {
    convertedTree.version = "007";
  }
  return {
    tree: convertedTree,
    isValid: errors.length === 0,
    isValidSearch: isValidSearch ?? sortedTreeErrors
  };
}
function convertFlatLezerTree({
  tree,
  numberParser,
  doc,
  internalValidatorMap,
  customTypes,
  parserConfig
}) {
  if (!doc || doc.toString() === "") {
    return { errors: [] };
  }
  let convertedTree = {
    type: "Group",
    children: [],
    range: { from: 0, to: tree.topNode.to },
    logicalOperator: "AND",
    explicitLogicalOperator: false,
    explicit: false
  };
  const errors = [];
  tree.iterate({
    from: tree.topNode.from,
    to: tree.topNode.to,
    enter: (_node) => {
      const node = _node.node;
      if (node.name === "Filter") {
        const { convertedNode, errors: groupErrors } = convertFilterGroup(
          node,
          doc,
          numberParser,
          internalValidatorMap,
          customTypes,
          parserConfig,
          "Filter"
        );
        convertedTree = convertedNode;
        errors.push(...groupErrors);
        return false;
      } else {
        console.log(
          `${doc.toString()} | Syntax node '${node.name}' cannot be converted. Please make sure to handle all nodes accordingly.`
        );
      }
    }
  });
  const filter = {
    tree: convertedTree,
    errors
  };
  return filter;
}
export {
  convertLezerTree,
  parserVersion
};
