// @dynatrace/strato-components/charts/xy-chart/components/selection/TooltipHeader.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TooltipHeader_exports = {};
__export(TooltipHeader_exports, {
  TooltipHeader: () => TooltipHeader
});
module.exports = __toCommonJS(TooltipHeader_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_units = require("@dynatrace-sdk/units");
var import_get_header_props = require("./get-header-props.js");
var import_useGridPlotSize = require("../../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_OverlayTooltip = require("../../../core/components/overlay-tooltip/OverlayTooltip.js");
var import_defaultUnitFormatter = require("../../../core/utils/defaultUnitFormatter.js");
var import_safe_get = require("../../../core/utils/safe-get.js");
var import_useAxesFormatter = require("../../hooks/useAxesFormatter.js");
var import_useVisibleAxesConfig = require("../../hooks/useVisibleAxesConfig.js");
var import_useVisibleAxesScales = require("../../hooks/useVisibleAxesScales.js");
var import_selectors = require("../../store/selectors.js");
function getAxisFormatter(axesIdToFormatter, axisConfig) {
  if (!axisConfig) {
    return void 0;
  }
  if (axisConfig.isRelative) {
    return (0, import_defaultUnitFormatter.defaultUnitFormatter)(import_units.units.percentage.percent);
  }
  return (0, import_safe_get.safeGet)(axesIdToFormatter, axisConfig.id);
}
const TooltipHeader = (props) => {
  const selectState = (0, import_selectors.useSelectState)();
  const axesIdToFormatter = (0, import_useAxesFormatter.useAxesFormatter)();
  const { top: topScale, bottom: bottomScale } = (0, import_useVisibleAxesScales.useVisibleAxesScales)();
  const { top: topConfig, bottom: bottomConfig } = (0, import_useVisibleAxesConfig.useVisibleAxesConfig)();
  const { width } = (0, import_useGridPlotSize.useGridPlotSize)();
  if (!selectState) {
    return null;
  }
  const topFormatter = getAxisFormatter(axesIdToFormatter, topConfig);
  const bottomFormatter = getAxisFormatter(axesIdToFormatter, bottomConfig);
  const start = selectState.selectionPercentage ? width * selectState.selectionPercentage[0] / 100 : selectState.start.relative.x;
  const end = selectState.selectionPercentage ? width * selectState.selectionPercentage[1] / 100 : selectState.end.relative.x;
  const topDuration = topConfig && topScale && !!topFormatter;
  const bottomDuration = bottomConfig && bottomScale && !!bottomFormatter;
  const header1Props = topDuration ? (0, import_get_header_props.getHeaderProps)({
    axisConfig: topConfig,
    scale: topScale,
    formatter: topFormatter,
    startPixels: start,
    endPixels: end,
    suffix: "1"
  }) : {};
  const header2Props = bottomDuration ? (0, import_get_header_props.getHeaderProps)({
    axisConfig: bottomConfig,
    scale: bottomScale,
    formatter: bottomFormatter,
    startPixels: start,
    endPixels: end,
    suffix: "2"
  }) : {};
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_OverlayTooltip.OverlayTooltip.Header,
    {
      ...header1Props,
      ...header2Props,
      subheader: props.subheader
    }
  );
};
