// @dynatrace/strato-components/charts/xy-chart/components/selection/XYChartSelectionArea.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var XYChartSelectionArea_exports = {};
__export(XYChartSelectionArea_exports, {
  XYChartSelectionArea: () => XYChartSelectionArea
});
module.exports = __toCommonJS(XYChartSelectionArea_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_XYChartSelectionAreaTooltip = require("./XYChartSelectionAreaTooltip.js");
var import_useGridPlotSize = require("../../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_OverlayContext = require("../../../core/components/overlay-tooltip/contexts/OverlayContext.js");
var import_DynamicSelectionArea = require("../../../core/components/selection-area/DynamicSelectionArea.js");
var import_useSynchronizedSelection = require("../../../core/components/shared-chart-interactions/hooks/useSynchronizedSelection.js");
var import_get_percentage_from_pixel = require("../../../core/utils/get-percentage-from-pixel.js");
var import_get_selection_pixels = require("../../../core/utils/get-selection-pixels.js");
var import_scale = require("../../../core/utils/scale.js");
var import_useAreaData = require("../../hooks/useAreaData.js");
var import_useAxesFormatter = require("../../hooks/useAxesFormatter.js");
var import_useBarData = require("../../hooks/useBarData.js");
var import_useDotData = require("../../hooks/useDotData.js");
var import_useFormatCategory = require("../../hooks/useFormatCategory.js");
var import_useLineData = require("../../hooks/useLineData.js");
var import_useRectData = require("../../hooks/useRectData.js");
var import_useSyncedXScaleByPriority = require("../../hooks/useSyncedXScaleByPriority.js");
var import_useVisibleAxesConfig = require("../../hooks/useVisibleAxesConfig.js");
var import_useVisibleAxesScales = require("../../hooks/useVisibleAxesScales.js");
var import_useXYChartData = require("../../hooks/useXYChartData.js");
var import_useXYChartId = require("../../hooks/useXYChartId.js");
var import_useXYChartScales = require("../../hooks/useXYChartScales.js");
var import_useXYChartSelectionAreaRef = require("../../hooks/useXYChartSelectionAreaRef.js");
var import_reducer_explorer = require("../../store/reducer-explorer.js");
var import_selectors = require("../../store/selectors.js");
var import_guards = require("../../types/guards.js");
var import_count_datapoints_with_series_and_categories = require("../../utils/count-datapoints-with-series-and-categories.js");
var import_get_selected_datapoints = require("../overlay/utils/get-selected-datapoints.js");
var import_handleFormatter = require("../overlay/utils/handleFormatter.js");
const XYChartSelectionArea = () => {
  const setState = (0, import_selectors.useSetState)();
  const { overlayRef: chartRef } = (0, import_OverlayContext.useOverlay)();
  const axesConfig = (0, import_useVisibleAxesConfig.useVisibleAxesConfig)();
  const axesIdToFormatter = (0, import_useAxesFormatter.useAxesFormatter)();
  const axesScales = (0, import_useVisibleAxesScales.useVisibleAxesScales)();
  const selectState = (0, import_selectors.useSelectState)();
  const { width } = (0, import_useGridPlotSize.useGridPlotSize)();
  const selectionAreaRef = (0, import_useXYChartSelectionAreaRef.useXYChartSelectionAreaRef)();
  const {
    filteredSeriesWithDatapoints,
    axes,
    axesMetadata,
    datapointMetadataMap,
    axisIdToScaleDomain
  } = (0, import_useXYChartData.useXYChartData)();
  const axisIdToScales = (0, import_useXYChartScales.useXYChartScales)();
  const mode = (0, import_selectors.useInteractionModeState)();
  const { updateSelectionState } = (0, import_useSynchronizedSelection.useSynchronizedSelection)();
  const chartId = (0, import_useXYChartId.useXYChartId)();
  const syncedXScale = (0, import_useSyncedXScaleByPriority.useSyncedXScaleByPriority)();
  const { seriesToMetadata: dotSeriesToMetadata } = (0, import_useDotData.useDotData)();
  const { seriesToMetadata: lineSeriesToMetadata } = (0, import_useLineData.useLineData)();
  const { seriesToMetadata: rectSeriesToMetadata } = (0, import_useRectData.useRectData)();
  const { seriesToMetadata: barSeriesToMetadata } = (0, import_useBarData.useBarData)();
  const { seriesToMetadata: areaSeriesToMetadata } = (0, import_useAreaData.useAreaData)();
  const formatCategory = (0, import_useFormatCategory.useFormatCategory)();
  const canZoom = axesMetadata.isZoomEnabled;
  if ((0, import_lodash_es.isUndefined)(selectState) || !chartRef.current) {
    return null;
  }
  const isBottomLabelRotated = axesConfig["bottom"]?.tickLabelLayout === "vertical";
  const isTopLabelRotated = axesConfig["top"]?.tickLabelLayout === "vertical";
  const { start, end, selectionPercentage, finished, accessible, variant } = selectState;
  const handleWrapperKeyDown = (event) => {
    if (event.key === "Enter" && canZoom) {
      chartRef.current?.focus();
    }
  };
  const handleSelectionChange = ({
    selection,
    accessible: accessible2
  }) => {
    setState(
      (0, import_reducer_explorer.reduceExploreSelect)({
        start: selection[0],
        end: selection[1],
        finished: true,
        accessible: accessible2,
        variant,
        selectionPercentage: (0, import_get_percentage_from_pixel.getPercentageFromPixel)(selection, width)
      })
    );
    if (mode === "explore" && syncedXScale) {
      updateSelectionState?.({
        selection,
        chartId,
        scale: syncedXScale
      });
    }
  };
  const formatterHandler = (config, scale) => {
    const axisConfig = {
      axisType: config?.type,
      axisScale: scale
    };
    if (config && (0, import_guards.isAnyNumericalScale)(axisConfig)) {
      return (value) => (0, import_handleFormatter.handleFormatter)(
        config,
        axesIdToFormatter,
        axisConfig.axisScale,
        value,
        width
      );
    } else if ((0, import_guards.isCategoricalScale)(axisConfig)) {
      return (value) => {
        const raw = (0, import_scale.scaleBandInvert)(axisConfig.axisScale)(value);
        return (raw && formatCategory(config, raw)) ?? raw;
      };
    }
  };
  const selectionStart = selectState.selectionPercentage ? width * selectState.selectionPercentage[0] / 100 : start.relative.x;
  const selectionEnd = selectState.selectionPercentage ? width * selectState.selectionPercentage[1] / 100 : end.relative.x;
  const selectedDatapointsPerSeries = (0, import_get_selected_datapoints.getSelectedDatapoints)(
    filteredSeriesWithDatapoints,
    axisIdToScales,
    axes,
    {
      start: selectionStart,
      end: selectionEnd
    },
    dotSeriesToMetadata,
    lineSeriesToMetadata,
    rectSeriesToMetadata,
    datapointMetadataMap,
    axisIdToScaleDomain,
    barSeriesToMetadata,
    areaSeriesToMetadata
  );
  const selectionPosition = finished ? (0, import_get_selection_pixels.getSelectionPixels)(selectionPercentage, width, [start, end]) : [start, end];
  const selectedDatapoints = selectedDatapointsPerSeries.map(
    (data) => data.datapoints
  );
  const datapointCounts = (0, import_count_datapoints_with_series_and_categories.countDatapointsWithSeriesAndCategories)(
    selectedDatapoints,
    filteredSeriesWithDatapoints,
    axes
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    variant === "select" ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_XYChartSelectionAreaTooltip.XYChartSelectionAreaTooltip,
      {
        ...datapointCounts,
        selectedDatapointsPerSeries
      }
    ) : null,
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_DynamicSelectionArea.DynamicSelectionArea,
      {
        selection: selectionPosition,
        chartRef,
        variant,
        focusable: accessible,
        interactable: finished,
        bottomFormatter: formatterHandler(axesConfig.bottom, axesScales.bottom),
        topFormatter: formatterHandler(axesConfig.top, axesScales.top),
        onKeyDown: handleWrapperKeyDown,
        onSelectionChange: handleSelectionChange,
        isBottomLabelRotated,
        isTopLabelRotated,
        selectionAreaRef,
        hasMagnifiers: !(0, import_lodash_es.isUndefined)(axesConfig.bottom) || !(0, import_lodash_es.isUndefined)(axesConfig.top)
      }
    )
  ] });
};
