// @dynatrace/strato-components/charts/xy-chart/components/selection/XYChartSelectionAreaTooltip.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var XYChartSelectionAreaTooltip_exports = {};
__export(XYChartSelectionAreaTooltip_exports, {
  XYChartSelectionAreaTooltip: () => XYChartSelectionAreaTooltip
});
module.exports = __toCommonJS(XYChartSelectionAreaTooltip_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_TooltipHeader = require("./TooltipHeader.js");
var import_get_custom_context_menu_items = require("../../../core/components/interactions-overlay/utils/get-custom-context-menu-items.js");
var import_OverlayContext = require("../../../core/components/overlay-tooltip/contexts/OverlayContext.js");
var import_OverlayTooltip = require("../../../core/components/overlay-tooltip/OverlayTooltip.js");
var import_SelectionTooltipWithAsyncActions = require("../../../core/components/selection-area/SelectionTooltipWithAsyncActions.js");
var import_get_selection_tooltip_position = require("../../../core/components/selection-area/tooltip/get-selection-tooltip-position.js");
var import_usePointerAndScrollResponsiveChartBounds = require("../../../core/hooks/usePointerAndScrollResponsiveChartBounds.js");
var import_useSelectConfig = require("../../hooks/useSelectConfig.js");
var import_useXYChartData = require("../../hooks/useXYChartData.js");
var import_messages = require("../../messages/messages.js");
var import_selectors = require("../../store/selectors.js");
var import_get_datapoints_message = require("../../utils/get-datapoints-message.js");
var import_useZoomToSelection = require("../overlay/hooks/useZoomToSelection.js");
const XYChartSelectionAreaTooltip = ({
  selectedDatapoints,
  totalSeries: seriesCount,
  totalCategories: categoriesCount,
  selectedDatapointsPerSeries
}) => {
  const selectState = (0, import_selectors.useSelectState)();
  const { axesMetadata, seriesWithDatapoints } = (0, import_useXYChartData.useXYChartData)();
  const selectConfig = (0, import_useSelectConfig.useSelectConfig)();
  if ((0, import_lodash_es.isUndefined)(selectState)) {
    return null;
  }
  const selectedDatapointsWithData = selectedDatapointsPerSeries.filter(
    (data) => data.datapoints.length
  );
  const selectedSeries = selectedDatapointsWithData.map((data) => {
    return {
      [data.seriesName]: data.datapoints
    };
  });
  const selectionDomain = selectedDatapointsWithData.map((data) => {
    return {
      [data.seriesName]: data.domain
    };
  });
  const datapointsMessage = (0, import_get_datapoints_message.getDatapointsMessage)(
    selectedDatapoints,
    seriesCount,
    seriesWithDatapoints.length,
    categoriesCount
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    DynamicSelectionAreaTooltip,
    {
      selectedDatapoints,
      selectState,
      categoriesCount,
      seriesCount,
      axesMetadata,
      selectedSeries,
      selectionDomain,
      selectConfig,
      datapointsMessage
    }
  );
};
const DynamicSelectionAreaTooltip = ({
  selectedDatapoints,
  selectState,
  categoriesCount,
  seriesCount,
  axesMetadata,
  selectedSeries,
  selectionDomain,
  selectConfig,
  datapointsMessage
}) => {
  const intl = (0, import_react_intl.useIntl)();
  const zoomToSelection = (0, import_useZoomToSelection.useZoomToSelection)();
  const { overlayRef: chartRef } = (0, import_OverlayContext.useOverlay)();
  const chartBounds = (0, import_usePointerAndScrollResponsiveChartBounds.usePointerAndScrollResponsiveChartBounds)(
    chartRef,
    true
  )();
  if ((0, import_lodash_es.isUndefined)(selectState) || !chartBounds) {
    return null;
  }
  const tooltipPosition = (0, import_get_selection_tooltip_position.getSelectionTooltipPosition)(selectState, chartBounds);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_SelectionTooltipWithAsyncActions.SelectionTooltipWithAsyncActions,
    {
      actions: () => {
        const actions = selectConfig?.actions;
        return actions && (async () => actions(selectedSeries, selectionDomain))().then(
          import_get_custom_context_menu_items.getCustomContextMenuItems
        );
      },
      selectFinished: selectState.finished,
      header: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_TooltipHeader.TooltipHeader,
        {
          subheader: intl.formatMessage(datapointsMessage, {
            count: selectedDatapoints,
            seriesCount: seriesCount + categoriesCount
          })
        }
      ),
      position: tooltipPosition,
      zoomEnabled: axesMetadata.isZoomEnabled,
      range: selectState.selectionPercentage || [0, 0],
      children: axesMetadata.isZoomEnabled ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_OverlayTooltip.OverlayTooltip.Action,
        {
          label: intl.formatMessage(import_messages.XYCHART_MESSAGES.zoomToSelection),
          icon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ZoomToSelectionIcon, {}),
          onSelect: zoomToSelection,
          trailingIcon: "\u23CE"
        }
      ) : null
    }
  );
};
