// @dynatrace/strato-components/tables/DataTable/features/ColumnPinning/ColumnPinning.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ColumnPinning_exports = {};
__export(ColumnPinning_exports, {
  DataTableColumnPinning: () => DataTableColumnPinning,
  useColumnPinning: () => useColumnPinning
});
module.exports = __toCommonJS(ColumnPinning_exports);
var import_react_table = require("@tanstack/react-table");
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_column_pinning_utils = require("./column-pinning-utils.js");
var import_useControllableState = require("../../../../core/hooks/useControllableState.js");
var import_constants = require("../../constants.js");
var import_datatable_typeguards = require("../../hooks/useTable/datatable-typeguards.js");
var import_utils = require("../../hooks/useTable/utils.js");
var import_layout_utils = require("../Layout/layout-utils.js");
function computeLeftPinnedPosition(columnId, table) {
  const leftVisibleLeafColumns = table.getLeftVisibleLeafColumns();
  const targetColumnIndex = leftVisibleLeafColumns.findIndex(
    (leaf) => leaf.id === columnId
  );
  if (targetColumnIndex < 0) {
    return void 0;
  }
  if (targetColumnIndex === 0) {
    return 0;
  }
  const prevColumn = leftVisibleLeafColumns[targetColumnIndex - 1];
  const isPrefix = prevColumn.columnDef.meta?.builtinPlacement === "prefix";
  const width = isPrefix ? (
    // prefix columns have a fixed width
    // use that instead of measured header, as prefix columns are also pinned when headers are hidden
    (0, import_layout_utils.getPrefixColumnWidth)(
      prevColumn,
      table.getVisualVariantOptions().rowDensity,
      table.options.rowThresholds
    )
  ) : table.getColumnOrHeaderWidth(prevColumn.id);
  const prevPosition = prevColumn.getLeftPinnedPosition() ?? 0;
  return prevPosition + width;
}
function computeRightPinnedPosition(columnId, table) {
  const rightVisibleLeafColumns = table.getRightVisibleLeafColumns();
  const targetColumnIndex = rightVisibleLeafColumns.findIndex(
    (leaf) => leaf.id === columnId
  );
  if (targetColumnIndex < 0) {
    return void 0;
  }
  if (targetColumnIndex === rightVisibleLeafColumns.length - 1) {
    return 0;
  }
  const nextColumn = rightVisibleLeafColumns[targetColumnIndex + 1];
  const nextPosition = nextColumn.getRightPinnedPosition() ?? 0;
  return nextPosition + table.getColumnOrHeaderWidth(nextColumn.id);
}
const DataTableColumnPinning = {
  createTable(table) {
    table.getOrderedHeaderGroups = (0, import_react_table.memo)(
      () => {
        const state = table.getState();
        return [
          table.options.columns,
          state.columnVisibility,
          state.columnPinning,
          state.columnOrder
        ];
      },
      () => {
        const leftGroups = table.getLeftHeaderGroups();
        const centerGroups = table.getCenterHeaderGroups();
        const rightGroups = table.getRightHeaderGroups();
        const getOrderedHeaderGroup = (groupIndex) => [
          ...leftGroups.at(groupIndex)?.headers ?? [],
          ...centerGroups.at(groupIndex)?.headers ?? [],
          ...rightGroups.at(groupIndex)?.headers ?? []
        ];
        return [getOrderedHeaderGroup(0), getOrderedHeaderGroup(1)];
      },
      (0, import_utils.getMemoOptions)(
        table.options,
        "debugColumns",
        "column.getOrderedHeaderGroups"
      )
    );
  },
  createHeader(header, table) {
    header.getIsPinned = () => {
      if ((0, import_datatable_typeguards.isGroupColumnDef)(header.column.columnDef)) {
        const firstChild = header.subHeaders.at(0);
        return firstChild?.getIsPinned() ?? false;
      }
      return header.column.getIsPinned();
    };
    header.getLeftPinnedPosition = () => {
      if ((0, import_datatable_typeguards.isGroupColumnDef)(header.column.columnDef)) {
        const firstChild = header.subHeaders.at(0);
        return firstChild?.column.getLeftPinnedPosition();
      }
      return header.column.getLeftPinnedPosition();
    };
    header.getRightPinnedPosition = () => {
      if ((0, import_datatable_typeguards.isGroupColumnDef)(header.column.columnDef)) {
        const lastChild = header.subHeaders.at(-1);
        return lastChild?.column.getRightPinnedPosition();
      }
      return header.column.getRightPinnedPosition();
    };
    header.getIsLastVisibleLeftPinnedHeader = () => {
      if ((0, import_datatable_typeguards.isGroupColumnDef)(header.column.columnDef)) {
        const lastChild = header.subHeaders.at(-1);
        return lastChild?.column.getIsLastVisibleLeftPinnedColumn() ?? false;
      }
      return header.column.getIsLastVisibleLeftPinnedColumn();
    };
    header.getIsFirstVisibleUnpinnedHeader = () => {
      if ((0, import_datatable_typeguards.isGroupColumnDef)(header.column.columnDef)) {
        const firstChild = header.subHeaders.at(0);
        return firstChild?.column.getIsFirstVisibleUnpinnedColumn() ?? false;
      }
      return header.column.getIsFirstVisibleUnpinnedColumn();
    };
    header.getIsLastVisibleUnpinnedHeader = () => {
      if ((0, import_datatable_typeguards.isGroupColumnDef)(header.column.columnDef)) {
        const lastChild = header.subHeaders.at(-1);
        return lastChild?.column.getIsLastVisibleUnpinnedColumn() ?? false;
      }
      return header.column.getIsLastVisibleUnpinnedColumn();
    };
    header.getIsFirstVisibleRightPinnedHeader = () => {
      if ((0, import_datatable_typeguards.isGroupColumnDef)(header.column.columnDef)) {
        const firstChild = header.subHeaders.at(0);
        return firstChild?.column.getIsFirstVisibleRightPinnedColumn() ?? false;
      }
      return header.column.getIsFirstVisibleRightPinnedColumn();
    };
    header.pin = (position) => {
      const { column, subHeaders: visibleSubHeaders } = header;
      const idsToPin = (0, import_datatable_typeguards.isGroupColumnDef)(column.columnDef) ? visibleSubHeaders.map((subHeader) => subHeader.column.id) : [column.id];
      const pinningState = (0, import_column_pinning_utils.pinColumns)(
        idsToPin,
        table.getState().columnPinning,
        position
      );
      table.setColumnPinning(pinningState);
    };
  },
  createColumn(column, table) {
    let cachedLeftPinnedPosition;
    column.getLeftPinnedPosition = (0, import_react_table.memo)(
      () => {
        const state = table.getState();
        return [
          state.columnSizingInfo.isResizingColumn,
          state.columnVisibility,
          state.columnPinning.left?.map((id) => `${id}:${table.getMeasuredColumnWidth(id)}`).join(),
          table.options.columns
        ];
      },
      (isResizingColumn) => {
        if (isResizingColumn) {
          if (cachedLeftPinnedPosition === void 0) {
            cachedLeftPinnedPosition = computeLeftPinnedPosition(
              column.id,
              table
            );
          }
          return cachedLeftPinnedPosition;
        }
        cachedLeftPinnedPosition = void 0;
        return computeLeftPinnedPosition(column.id, table);
      },
      (0, import_utils.getMemoOptions)(
        table.options,
        "debugColumns",
        "column.getLeftPinnedPosition"
      )
    );
    let cachedRightPinnedPosition;
    column.getRightPinnedPosition = (0, import_react_table.memo)(
      () => {
        const state = table.getState();
        return [
          state.columnSizingInfo.isResizingColumn,
          state.columnVisibility,
          state.columnPinning.right?.map((id) => `${id}:${table.getMeasuredColumnWidth(id)}`).join(),
          table.options.columns
        ];
      },
      (isResizingColumn) => {
        if (isResizingColumn) {
          if (cachedRightPinnedPosition === void 0) {
            cachedRightPinnedPosition = computeRightPinnedPosition(
              column.id,
              table
            );
          }
          return cachedRightPinnedPosition;
        }
        cachedRightPinnedPosition = void 0;
        return computeRightPinnedPosition(column.id, table);
      },
      (0, import_utils.getMemoOptions)(
        table.options,
        "debugColumns",
        "column.getRightPinnedPosition"
      )
    );
    column.getIsLastVisibleLeftPinnedColumn = (0, import_react_table.memo)(
      () => {
        const state = table.getState();
        return [
          state.columnPinning.left,
          state.columnVisibility,
          table.options.columns
        ];
      },
      () => {
        const lastLeftVisibleLeafColumnId = table.getLeftVisibleLeafColumns().at(-1)?.id;
        return column.id === lastLeftVisibleLeafColumnId || column.columns.some(
          (childColumn) => childColumn.id === lastLeftVisibleLeafColumnId
        );
      },
      (0, import_utils.getMemoOptions)(
        table.options,
        "debugColumns",
        "column.getIsLastVisibleLeftPinnedColumn"
      )
    );
    column.getIsFirstVisibleUnpinnedColumn = (0, import_react_table.memo)(
      () => {
        const state = table.getState();
        return [
          state.columnPinning,
          state.columnVisibility,
          state.columnOrder,
          table.options.columns
        ];
      },
      () => {
        const firstCenterVisibleLeafColumnId = table.getCenterVisibleLeafColumns().at(0)?.id;
        return column.id === firstCenterVisibleLeafColumnId || column.columns.some(
          (childColumn) => childColumn.id === firstCenterVisibleLeafColumnId
        );
      },
      (0, import_utils.getMemoOptions)(
        table.options,
        "debugColumns",
        "column.getIsFirstVisibleUnpinnedColumn"
      )
    );
    column.getIsLastVisibleUnpinnedColumn = (0, import_react_table.memo)(
      () => {
        const state = table.getState();
        return [
          state.columnPinning,
          state.columnVisibility,
          state.columnOrder,
          table.options.columns
        ];
      },
      () => {
        const lastCenterVisibleLeafColumnId = table.getCenterVisibleLeafColumns().at(-1)?.id;
        return column.id === lastCenterVisibleLeafColumnId || column.columns.some(
          (childColumn) => childColumn.id === lastCenterVisibleLeafColumnId
        );
      },
      (0, import_utils.getMemoOptions)(
        table.options,
        "debugColumns",
        "column.getIsLastVisibleUnpinnedColumn"
      )
    );
    column.getIsFirstVisibleRightPinnedColumn = (0, import_react_table.memo)(
      () => {
        const state = table.getState();
        return [
          state.columnPinning.right,
          state.columnVisibility,
          table.options.columns
        ];
      },
      () => {
        const firstRightVisibleLeafColumnId = table.getRightVisibleLeafColumns().at(0)?.id;
        return column.id === firstRightVisibleLeafColumnId || column.columns.some(
          (childColumn) => childColumn.id === firstRightVisibleLeafColumnId
        );
      },
      (0, import_utils.getMemoOptions)(
        table.options,
        "debugColumns",
        "column.getIsFirstVisibleRightPinnedColumn"
      )
    );
  }
};
function isCustomColumn(columnId) {
  return !import_constants.DATA_TABLE_BUILT_IN_COLUMN_IDS.has(columnId);
}
function removeBuiltinColumns(pinnedColumns) {
  return pinnedColumns?.filter(isCustomColumn) ?? [];
}
function addBuiltinColumns(pinnedColumns, options) {
  if (!pinnedColumns) {
    return {};
  }
  const leftPinnedColumns = removeBuiltinColumns(pinnedColumns.left);
  const rightPinnedColumns = removeBuiltinColumns(pinnedColumns.right);
  if (options.enableRowDetails) {
    leftPinnedColumns.unshift(import_constants.DATA_TABLE_ROW_DETAILS_COLUMN_ID);
  }
  if (options.enableRowSelection) {
    leftPinnedColumns.unshift(import_constants.DATA_TABLE_ROW_SELECTION_COLUMN_ID);
  }
  if (options.enableRowOrderDragAndDrop) {
    leftPinnedColumns.unshift(import_constants.DATA_TABLE_ROW_DRAG_COLUMN_ID);
  }
  if (options.enableRowActions) {
    rightPinnedColumns.push(import_constants.DATA_TABLE_ROW_ACTIONS_COLUMN_ID);
  }
  return {
    left: leftPinnedColumns,
    right: rightPinnedColumns
  };
}
function useColumnPinning(props, options) {
  options.enableColumnPinning = props.columnPinning === true;
  const {
    enableColumnPinning,
    enableRowDetails,
    enableRowSelection,
    enableRowOrderDragAndDrop,
    enableRowActions
  } = options;
  const defaultPinnedColumns = (0, import_react.useMemo)(() => {
    const userDefinedPinnedColumns = enableColumnPinning && props.defaultPinnedColumns ? props.defaultPinnedColumns : {};
    return addBuiltinColumns(userDefinedPinnedColumns, {
      enableRowDetails: !!enableRowDetails,
      enableRowSelection: !!enableRowSelection,
      enableRowOrderDragAndDrop: !!enableRowOrderDragAndDrop,
      enableRowActions: !!enableRowActions
    });
  }, [
    enableColumnPinning,
    enableRowActions,
    enableRowDetails,
    enableRowOrderDragAndDrop,
    enableRowSelection,
    props.defaultPinnedColumns
  ]);
  const pinnedColumns = (0, import_react.useMemo)(
    () => enableColumnPinning && props.pinnedColumns ? addBuiltinColumns(props.pinnedColumns, {
      enableRowDetails: !!enableRowDetails,
      enableRowSelection: !!enableRowSelection,
      enableRowOrderDragAndDrop: !!enableRowOrderDragAndDrop,
      enableRowActions: !!enableRowActions
    }) : void 0,
    [
      enableColumnPinning,
      enableRowActions,
      enableRowDetails,
      enableRowOrderDragAndDrop,
      enableRowSelection,
      props.pinnedColumns
    ]
  );
  const [columnPinning, setColumnPinning] = (0, import_useControllableState.useControllableState)({
    value: pinnedColumns,
    defaultValue: defaultPinnedColumns
  });
  options.state = {
    ...options.state,
    columnPinning: enableColumnPinning ? columnPinning : defaultPinnedColumns
    // just built-in columns
  };
  options.onColumnPinningChange = (0, import_react.useCallback)(
    (updater) => {
      const nextState = (0, import_react_table.functionalUpdate)(updater, columnPinning);
      if ((0, import_lodash_es.isEqual)(nextState, columnPinning)) {
        return;
      }
      setColumnPinning(
        addBuiltinColumns(nextState, {
          enableRowDetails: !!enableRowDetails,
          enableRowSelection: !!enableRowSelection,
          enableRowOrderDragAndDrop: !!enableRowOrderDragAndDrop,
          enableRowActions: !!enableRowActions
        })
      );
      props.onPinnedColumnsChange?.({
        left: removeBuiltinColumns(nextState.left),
        right: removeBuiltinColumns(nextState.right)
      });
    },
    [
      columnPinning,
      enableRowActions,
      enableRowDetails,
      enableRowOrderDragAndDrop,
      enableRowSelection,
      props,
      setColumnPinning
    ]
  );
  const originalColumnPinningRef = (0, import_react.useRef)();
  if (!originalColumnPinningRef.current) {
    const originalPinned = pinnedColumns || defaultPinnedColumns || [];
    originalColumnPinningRef.current = {
      left: originalPinned.left,
      right: originalPinned.right
    };
  }
  options.originalColumnPinning = originalColumnPinningRef.current;
}
