// @dynatrace/strato-components/esm/charts/core/components/sparkline-display/utils/accessors.js
import { isIsolatedPoint } from "../../../utils/is-isolated-point.js";
import { sanitizeZeroValues } from "../../../utils/sanitize-zero-values.js";
import {
  TIMESTAMP_BAR_WIDTH,
  TIMESTAMP_BAR_X_DEVIATION
} from "../constants.js";
const xAxisAccessor = (xScale) => (dataPoint) => xScale(dataPoint.end ?? dataPoint.start) ?? 0;
const yAccessor = (yScale) => (dataPoint) => yScale(dataPoint.value) ?? 0;
const y0Accessor = (yScale) => () => sanitizeZeroValues(yScale, 0) ?? 0;
const y1Accessor = (yScale) => (dataPoint) => yScale(dataPoint.value) ?? 0;
const definedAccessor = (datapoint) => {
  return !datapoint.isGap;
};
const isIsolatedAccessor = (dataPoints, index) => isIsolatedPoint(dataPoints, index, (point) => definedAccessor(point));
const barXAxisAccessor = (xScale) => (dataPoint) => {
  const { start, end } = dataPoint;
  const isTimeframe = end && end !== start;
  const deviation = isTimeframe ? 0 : TIMESTAMP_BAR_X_DEVIATION;
  return (xScale(start) ?? 0) - deviation;
};
const barYAxisAccessor = (yScale) => (dataPoint) => {
  const maxValue = yScale.invert(0);
  const maxValueCoordinate = sanitizeZeroValues(yScale, maxValue);
  const cumulativeCoordinate = sanitizeZeroValues(yScale, dataPoint.value);
  return maxValueCoordinate + cumulativeCoordinate;
};
const barWidthAccessor = (datapoint, xScale) => {
  const { start, end } = datapoint;
  const isTimeframe = end && end !== start;
  return isTimeframe ? (xScale(end) ?? 0) - (xScale(start) ?? 0) : TIMESTAMP_BAR_WIDTH;
};
const barHeightAccessor = (datapoint, yScale) => {
  const valueCoordinate = sanitizeZeroValues(yScale, datapoint.value);
  const zeroCoordinate = sanitizeZeroValues(yScale, 0);
  return zeroCoordinate - valueCoordinate;
};
export {
  barHeightAccessor,
  barWidthAccessor,
  barXAxisAccessor,
  barYAxisAccessor,
  definedAccessor,
  isIsolatedAccessor,
  xAxisAccessor,
  y0Accessor,
  y1Accessor,
  yAccessor
};
