// @dynatrace/strato-components/esm/charts/core/components/sparkline-display/utils/sanitize-sparkline-data.js
import { isNil } from "lodash-es";
import { getDatapointIsGap } from "../../../utils/get-data-point-is-gap.js";
function isInXAxisBounds(datapoint, xAxis, customMin, customMax) {
  if (customMin && datapoint.start < xAxis.min) {
    return false;
  }
  return !(customMax && datapoint.start > xAxis.max);
}
function findNextDefinedIndex(datapoints, startIndex) {
  let i = startIndex;
  while (i < datapoints.length && isNil(datapoints[i].value)) {
    i++;
  }
  return i;
}
function updateXAxisBounds(datapoint, xAxisBounds, customMin, customMax) {
  if (customMin && datapoint.start < xAxisBounds.min) {
    xAxisBounds.min = datapoint.start;
  }
  if (customMax) {
    const endDate = datapoint.end ?? datapoint.start;
    if (endDate > xAxisBounds.max) {
      xAxisBounds.max = endDate;
    }
  }
}
function processGap(datapoints, currentIndex, currentDatapoint, sanitizedDatapoints) {
  const nextIndex = findNextDefinedIndex(datapoints, currentIndex + 1);
  const nextDatapoint = datapoints[nextIndex];
  if (nextDatapoint) {
    const isGap = getDatapointIsGap(currentDatapoint, nextDatapoint);
    if (isGap) {
      sanitizedDatapoints.push({
        start: currentDatapoint.end ?? currentDatapoint.start,
        end: nextDatapoint.start,
        value: Infinity,
        isGap: true
      });
    }
  }
}
function createBoundaryDatapoint(date) {
  return { start: new Date(date), value: Infinity, isGap: true };
}
function addXAxisBoundaryDatapoints(sanitizedDatapoints, xAxis, xAxisBounds, customMin, customMax) {
  if (customMin && xAxisBounds.min > xAxis.min) {
    sanitizedDatapoints.unshift(createBoundaryDatapoint(xAxis.min));
  }
  if (customMax && xAxisBounds.max < xAxis.max) {
    sanitizedDatapoints.push(createBoundaryDatapoint(xAxis.max));
  }
}
const sanitizeSparklineData = (series, gapPolicy, xAxis) => {
  const domainValues = { min: Infinity, max: -Infinity };
  if (isNil(series) || !series.datapoints?.length) {
    return { sanitizedSeries: { datapoints: [], name: "" }, domainValues };
  }
  const customXAxisMin = xAxis.min !== "data-min";
  const customXAxisMax = xAxis.max !== "data-max";
  const sanitizedDatapoints = [];
  const xAxisBounds = {
    min: series.datapoints[0].start,
    max: series.datapoints[0].start
  };
  for (let i = 0; i < series.datapoints.length; i++) {
    const currentDatapoint = series.datapoints[i];
    if (isNil(currentDatapoint.value)) {
      continue;
    }
    if (!isInXAxisBounds(currentDatapoint, xAxis, customXAxisMin, customXAxisMax)) {
      continue;
    }
    domainValues.min = Math.min(domainValues.min, currentDatapoint.value);
    domainValues.max = Math.max(domainValues.max, currentDatapoint.value);
    updateXAxisBounds(
      currentDatapoint,
      xAxisBounds,
      customXAxisMin,
      customXAxisMax
    );
    sanitizedDatapoints.push(currentDatapoint);
    if (gapPolicy === "gap") {
      processGap(series.datapoints, i, currentDatapoint, sanitizedDatapoints);
    }
  }
  addXAxisBoundaryDatapoints(
    sanitizedDatapoints,
    xAxis,
    xAxisBounds,
    customXAxisMin,
    customXAxisMax
  );
  return {
    sanitizedSeries: { ...series, datapoints: sanitizedDatapoints },
    domainValues
  };
};
export {
  sanitizeSparklineData
};
