// @dynatrace/strato-components/tables/DataTable/components/Toolbar/ColumnSettings/column-settings-utils.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var column_settings_utils_exports = {};
__export(column_settings_utils_exports, {
  generateColumnLabelsMap: () => generateColumnLabelsMap,
  generateColumnOrderWithParentIds: () => generateColumnOrderWithParentIds,
  generateColumnVisibilityObject: () => generateColumnVisibilityObject,
  getCombinedVisibilityCheckboxState: () => getCombinedVisibilityCheckboxState,
  getHasColumnOrderChanged: () => getHasColumnOrderChanged,
  getHasColumnPinningChanged: () => getHasColumnPinningChanged,
  getHasVisibilityChanged: () => getHasVisibilityChanged,
  getIsSettingEnabled: () => getIsSettingEnabled,
  getMultipleLocalColumnVisibility: () => getMultipleLocalColumnVisibility,
  getResetOverride: () => getResetOverride,
  getVisibilityCheckboxState: () => getVisibilityCheckboxState,
  groupColumns: () => groupColumns,
  isParentColumnHideable: () => isParentColumnHideable,
  processFilterValue: () => processFilterValue
});
module.exports = __toCommonJS(column_settings_utils_exports);
var import_lodash_es = require("lodash");
var import_get_column_label = require("../../../../utils/get-column-label.js");
function generateColumnLabelsMap(columns) {
  const labels = /* @__PURE__ */ new Map();
  for (const col of columns) {
    labels.set(col.id, (0, import_get_column_label.getColumnLabel)(col).toLocaleLowerCase());
  }
  return labels;
}
function processFilterValue(filterValue) {
  return filterValue.toLocaleLowerCase().split(" ").filter(Boolean);
}
function getResetOverride(override, key) {
  return typeof override === "object" ? override[key] : void 0;
}
function getIsSettingEnabled(override, fallback) {
  return override === void 0 ? fallback : override !== false;
}
function groupColumns(columns) {
  const groupedColumns = [];
  let lastParent = void 0;
  for (const col of columns) {
    const parent = col.parent;
    if (parent === void 0) {
      lastParent = void 0;
      groupedColumns.push(col);
      continue;
    }
    if (parent !== lastParent) {
      groupedColumns.push(parent);
      lastParent = parent;
    }
    groupedColumns.push(col);
  }
  return groupedColumns;
}
function isParentColumnHideable(childColumns) {
  return childColumns.some((c) => c.columnDef.enableHiding);
}
function generateColumnVisibilityObject(columns, columnVisibility) {
  return Object.fromEntries(
    columns.map((column) => [
      column.id,
      columnVisibility ? columnVisibility[column.id] ?? true : column.getIsVisible()
    ])
  );
}
function getMultipleLocalColumnVisibility(columns, currentVisibility, localColumnsVisibilityState) {
  const newVisibility = currentVisibility !== true;
  const newState = { ...localColumnsVisibilityState };
  for (const column of columns) {
    if (column.getCanHide()) {
      newState[column.id] = newVisibility;
    }
  }
  return newState;
}
function getVisibilityCheckboxState(columnVisibilityState, column) {
  const childColumns = column.columns;
  if (childColumns.length === 0) {
    return columnVisibilityState?.[column.id] ?? false;
  }
  const childrenVisibility = generateColumnVisibilityObject(
    childColumns,
    columnVisibilityState
  );
  return getCombinedVisibilityCheckboxState(childrenVisibility, childColumns);
}
function getCombinedVisibilityCheckboxState(columnVisibilityState, columns) {
  let allVisible = true;
  let allHidden = true;
  for (const col of columns) {
    const isVisible = columnVisibilityState[col.id] !== false;
    if (!isVisible && col.getCanHide() === false) {
      continue;
    }
    if (isVisible) {
      allHidden = false;
    } else {
      allVisible = false;
    }
    if (!allVisible && !allHidden) {
      return "indeterminate";
    }
  }
  if (allVisible) {
    return true;
  }
  if (allHidden) {
    return false;
  }
  return "indeterminate";
}
function getHasVisibilityChanged(columnVisibilityToReset, localColumnVisibility) {
  if (!columnVisibilityToReset) {
    return false;
  }
  return Object.keys(localColumnVisibility).some((key) => {
    if (!(key in columnVisibilityToReset)) {
      return localColumnVisibility[key] !== true;
    }
    return localColumnVisibility[key] !== columnVisibilityToReset[key];
  });
}
function generateColumnOrderWithParentIds(getColumn, columnOrder) {
  const columnOrderWithParents = [];
  const uniqueParentIds = /* @__PURE__ */ new Set();
  for (const columnId of columnOrder) {
    const column = getColumn(columnId);
    const parentId = column?.parent?.id;
    if (parentId && !uniqueParentIds.has(parentId)) {
      columnOrderWithParents.push(parentId);
      uniqueParentIds.add(parentId);
    }
    columnOrderWithParents.push(columnId);
  }
  return {
    columnOrderWithParents,
    uniqueParentIds: Array.from(uniqueParentIds)
  };
}
function getHasColumnOrderChanged(columnOrderToReset, localColumnOrder) {
  if (!columnOrderToReset) {
    return false;
  }
  return !(0, import_lodash_es.isEqual)(columnOrderToReset, localColumnOrder);
}
function getHasColumnPinningChanged(columnPinnedToReset, localColumnsPinned) {
  if (!columnPinnedToReset) {
    return false;
  }
  return !(0, import_lodash_es.isEqual)(columnPinnedToReset.left, localColumnsPinned.left) || !(0, import_lodash_es.isEqual)(columnPinnedToReset.right, localColumnsPinned.right);
}
