// @dynatrace/strato-components/tables/DataTable/components/Toolbar/ColumnSettings/ColumnPinningMenu.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ColumnPinningMenu_exports = {};
__export(ColumnPinningMenu_exports, {
  ColumnPinningMenu: () => ColumnPinningMenu
});
module.exports = __toCommonJS(ColumnPinningMenu_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_Button = require("../../../../../buttons/button/Button.js");
var import_Menu = require("../../../../../navigation/menu/Menu.js");
var import_Tooltip = require("../../../../../overlays/tooltip/Tooltip.js");
var import_column_pinning_utils = require("../../../features/ColumnPinning/column-pinning-utils.js");
var import_messages = require("../../../messages.js");
const ColumnPinningMenu = (props) => {
  const {
    column,
    orderedFlatColumns,
    localColumnPinning,
    updateLocalColumnPinning
  } = props;
  const intl = (0, import_react_intl.useIntl)();
  const columnId = column.id;
  const pinnedPosition = column.pinned;
  const pinningLabel = intl.formatMessage(import_messages.dataTableMessages.pinColumnLabel, {
    isUnpinned: pinnedPosition === false
  });
  function getColumnsInSamePinningRegion() {
    const position = pinnedPosition === false ? "center" : pinnedPosition;
    return orderedFlatColumns[position];
  }
  function getIdsToPin() {
    const hasNestedColumns = column.columns.length;
    if (!hasNestedColumns) {
      return [columnId];
    }
    const idsToPin = [];
    const columns = getColumnsInSamePinningRegion();
    const startIndex = columns.findIndex((col) => col.key === column.key);
    if (startIndex < 0) {
      return [];
    }
    let index = startIndex + 1;
    let current = columns.at(index);
    while (current?.parent?.id === columnId) {
      idsToPin.push(current.id);
      index++;
      current = columns.at(index);
    }
    return idsToPin;
  }
  const pin = (position) => {
    const idsToPin = getIdsToPin();
    const newState = (0, import_column_pinning_utils.pinColumns)(idsToPin, localColumnPinning, position);
    updateLocalColumnPinning(newState);
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Tooltip.Tooltip, { text: pinningLabel, placement: "right", children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Menu.Menu, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Trigger, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button, { size: "condensed", "aria-label": pinningLabel, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.PinIcon, {}) }) }) }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Menu.Menu.Content, { children: [
      pinnedPosition !== "left" && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_Menu.Menu.Item,
        {
          onSelect: () => pin("left"),
          "data-dt-name": "ColumnSettingsPinLeft",
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_intl.FormattedMessage, { ...import_messages.dataTableMessages.pinColumnLeftLabel })
        }
      ),
      pinnedPosition !== "right" && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_Menu.Menu.Item,
        {
          onSelect: () => pin("right"),
          "data-dt-name": "ColumnSettingsPinRight",
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_intl.FormattedMessage, { ...import_messages.dataTableMessages.pinColumnRightLabel })
        }
      ),
      pinnedPosition !== false && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_Menu.Menu.Item,
        {
          onSelect: () => pin(false),
          "data-dt-name": "ColumnSettingsUnpin",
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_intl.FormattedMessage, { ...import_messages.dataTableMessages.unpinColumnLabel })
        }
      )
    ] })
  ] }) });
};
