// @dynatrace/strato-components/tables/DataTable/components/Toolbar/ColumnSettings/DataTableVisibilityFilterControls.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DataTableVisibilityFilterControls_exports = {};
__export(DataTableVisibilityFilterControls_exports, {
  DataTableVisibilityFilterControls: () => DataTableVisibilityFilterControls
});
module.exports = __toCommonJS(DataTableVisibilityFilterControls_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_react_intl = require("react-intl");
var styles = __toESM(require("./DataTableVisibilityFilterControls.sty.js"));
var import_Checkbox = require("../../../../../forms/checkbox/Checkbox.js");
var import_SearchInput = require("../../../../../forms/search-input/SearchInput.js");
var import_Flex = require("../../../../../layouts/flex/Flex.js");
var import_messages = require("../../../messages.js");
const SEARCH_DEBOUNCE_TIME_MS = 300;
const DataTableVisibilityFilterControls = (props) => {
  const {
    setActiveFilterValue,
    showHideAllState,
    onShowHideAllChange,
    showVisibilityControl,
    showFilterControl
  } = props;
  const intl = (0, import_react_intl.useIntl)();
  const [filterValue, setFilterValue] = (0, import_react.useState)("");
  const debouncedUpdateRef = (0, import_react.useRef)(
    (0, import_lodash_es.debounce)((value, setter) => {
      setter(value);
    }, SEARCH_DEBOUNCE_TIME_MS)
  );
  const onChange = (0, import_react.useCallback)(
    (value) => {
      setFilterValue(value);
      debouncedUpdateRef.current(value, setActiveFilterValue);
    },
    [setActiveFilterValue]
  );
  (0, import_react.useEffect)(() => {
    const debouncedUpdate = debouncedUpdateRef.current;
    return () => {
      debouncedUpdate.cancel();
    };
  }, []);
  if (!showFilterControl && !showVisibilityControl) {
    return null;
  }
  const dataDtProperties = `state:${showHideAllState === true ? "hide" : "show"}`;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_Flex.Flex,
    {
      justifyContent: "space-between",
      alignItems: "center",
      className: styles.filterControls,
      children: [
        showVisibilityControl && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_Checkbox.Checkbox,
          {
            value: showHideAllState,
            onChange: onShowHideAllChange,
            className: styles.showHideAll,
            "data-dt-name": "ColumnSettingsShowHideAll",
            "data-dt-properties": dataDtProperties,
            children: intl.formatMessage(import_messages.dataTableMessages.showHideAll)
          }
        ),
        showFilterControl && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_SearchInput.SearchInput,
          {
            placeholder: intl.formatMessage(
              import_messages.dataTableMessages.searchBarPlaceholder
            ),
            className: styles.searchInput,
            onChange,
            value: filterValue,
            "data-dt-name": "ColumnSettingsFilter"
          }
        )
      ]
    }
  );
};
