// @dynatrace/strato-components/tables/DataTable/components/Toolbar/ColumnSettings/DataTableColumnsListItem.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DataTableColumnsListItem_exports = {};
__export(DataTableColumnsListItem_exports, {
  DataTableColumnsListItem: () => DataTableColumnsListItem
});
module.exports = __toCommonJS(DataTableColumnsListItem_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_sortable = require("@dnd-kit/sortable");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_ColumnOrderDragHandle = require("./ColumnOrderSettings/ColumnOrderDragHandle.js");
var import_ColumnPinningMenu = require("./ColumnPinningMenu.js");
var import_ColumnVisibilityCheckbox = require("./ColumnVisibilityCheckbox.js");
var styles = __toESM(require("./DataTableColumnsListItem.sty.js"));
var import_useMergeRefs = require("../../../../../core/hooks/useMergeRefs.js");
var import_Flex = require("../../../../../layouts/flex/Flex.js");
var import_get_column_label = require("../../../../utils/get-column-label.js");
const InternalDataTableColumnsListItem = (props) => {
  const {
    column,
    columnIndex,
    orderedFlatColumns,
    showVisibilityCheckbox,
    enableVisibilityCheckbox,
    currentVisibilityState,
    onVisibilityChange,
    showColumnOrderDragHandle,
    showColumnPinningMenu,
    localColumnPinning,
    updateLocalColumnPinning,
    virtualIndex,
    measureElement
  } = props;
  const columnName = (0, import_get_column_label.getColumnLabel)(column);
  const localColumnRef = (0, import_react.useRef)(null);
  const { setNodeRef, isDragging } = (0, import_sortable.useSortable)({
    id: column.key,
    data: {
      column,
      columnId: column.id,
      columnIndex,
      localColumnRef
    }
  });
  const mergedRefs = (0, import_useMergeRefs.useMergeRefs)([localColumnRef, setNodeRef, measureElement]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_Flex.Flex,
    {
      "data-index": virtualIndex,
      "data-columnid": column.id,
      "data-parentcolumnid": column.parent?.id,
      "data-pinned-position": column.pinned === false ? void 0 : column.pinned,
      ref: mergedRefs,
      justifyContent: "space-between",
      alignItems: "center",
      className: (0, import_clsx.default)("strato-table-settings-item", styles.columnsListItem, {
        [styles.dragging]: isDragging
      }),
      style: {
        gridRowStart: virtualIndex + 1
      },
      children: [
        showVisibilityCheckbox ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_ColumnVisibilityCheckbox.ColumnVisibilityCheckbox,
          {
            column,
            enableCheckbox: enableVisibilityCheckbox,
            currentState: currentVisibilityState,
            onVisibilityChange
          }
        ) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: styles.columnName, children: columnName }),
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { gap: 4, children: [
          showColumnOrderDragHandle && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_ColumnOrderDragHandle.ColumnOrderDragHandle,
            {
              columnId: column.id,
              columnKey: column.key,
              isDraggingActive: isDragging
            }
          ),
          showColumnPinningMenu && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_ColumnPinningMenu.ColumnPinningMenu,
            {
              column,
              orderedFlatColumns,
              localColumnPinning,
              updateLocalColumnPinning
            }
          )
        ] })
      ]
    }
  );
};
const DataTableColumnsListItem = (0, import_react.memo)(
  InternalDataTableColumnsListItem,
  (prevProps, nextProps) => {
    return prevProps.column.key === nextProps.column.key && prevProps.columnIndex === nextProps.columnIndex && prevProps.showVisibilityCheckbox === nextProps.showVisibilityCheckbox && prevProps.enableVisibilityCheckbox === nextProps.enableVisibilityCheckbox && prevProps.currentVisibilityState === nextProps.currentVisibilityState && prevProps.showColumnOrderDragHandle === nextProps.showColumnOrderDragHandle && prevProps.showColumnPinningMenu === nextProps.showColumnPinningMenu && prevProps.localColumnPinning === nextProps.localColumnPinning && prevProps.virtualIndex === nextProps.virtualIndex;
  }
);
