// @dynatrace/strato-components/tables/DataTable/components/Toolbar/ColumnSettings/DataTableColumnSettingsModal.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DataTableColumnSettingsModal_exports = {};
__export(DataTableColumnSettingsModal_exports, {
  DataTableColumnSettingsModal: () => DataTableColumnSettingsModal
});
module.exports = __toCommonJS(DataTableColumnSettingsModal_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_column_settings_utils = require("./column-settings-utils.js");
var import_DataTableColumnsList = require("./DataTableColumnsList.js");
var import_DataTableVisibilityFilterControls = require("./DataTableVisibilityFilterControls.js");
var import_DataTableContext = require("../../../contexts/DataTableContext.js");
var import_messages = require("../../../messages.js");
var import_DataTableSettingsModal = require("../../SettingsModal/DataTableSettingsModal.js");
const ColumnSettingsModal = (props) => {
  const {
    visibilitySettings: toolbarVisibilitySettings,
    columnOrderSettings: toolbarColumnOrderSettings,
    columnPinningSettings: toolbarColumnPinningSettings,
    resetColumnOrder,
    resetColumnVisibility,
    resetColumnPinning,
    onModalClose,
    columnSettingsOverride
  } = props;
  const intl = (0, import_react_intl.useIntl)();
  const table = (0, import_DataTableContext.useDataTable)();
  const leafColumnsWithoutBuiltin = table.getLeafColumnsWithoutBuiltin();
  const visibilityOverride = columnSettingsOverride?.columnVisibility;
  const columnOrderOverride = columnSettingsOverride?.columnOrder;
  const columnPinningOverride = columnSettingsOverride?.columnPinning;
  const visibilitySettingsEnabled = (0, import_column_settings_utils.getIsSettingEnabled)(
    visibilityOverride,
    toolbarVisibilitySettings
  );
  const columnOrderSettingsEnabled = (0, import_column_settings_utils.getIsSettingEnabled)(
    columnOrderOverride,
    toolbarColumnOrderSettings
  );
  const columnPinningSettingsEnabled = (0, import_column_settings_utils.getIsSettingEnabled)(
    columnPinningOverride,
    toolbarColumnPinningSettings
  );
  const {
    columnOrderWithParents: tableColumnOrder,
    uniqueParentIds: allLocalColumnsParentIds
  } = (0, import_column_settings_utils.generateColumnOrderWithParentIds)(
    table.getColumn,
    table.getState().columnOrder
  );
  const [activeFilterValue, setActiveFilterValue] = (0, import_react.useState)("");
  const initialColumnVisibility = (0, import_react.useMemo)(
    () => (0, import_column_settings_utils.generateColumnVisibilityObject)(leafColumnsWithoutBuiltin),
    [leafColumnsWithoutBuiltin]
  );
  const [localColumnVisibility, setLocalColumnVisibility] = (0, import_react.useState)(initialColumnVisibility);
  const [localColumnOrder, setLocalColumnOrder] = (0, import_react.useState)(tableColumnOrder);
  const [localColumnPinning, setLocalColumnPinning] = (0, import_react.useState)(
    // original column pinning state
    () => table.getState().columnPinning
  );
  const [showResetStatus, setShowResetStatus] = (0, import_react.useState)(false);
  const columnOrderToReset = (0, import_column_settings_utils.getResetOverride)(columnOrderOverride, "resetColumnOrder") ?? resetColumnOrder ?? table.options.originalColumnOrder;
  const columnVisibilityToReset = (0, import_column_settings_utils.getResetOverride)(visibilityOverride, "resetColumnVisibility") ?? resetColumnVisibility ?? table.options.originalColumnVisibility;
  const columnPinningToReset = (0, import_column_settings_utils.getResetOverride)(columnPinningOverride, "resetColumnPinning") ?? resetColumnPinning ?? table.options.originalColumnPinning;
  const filteredLocalColumnOrder = (0, import_react.useMemo)(
    () => localColumnOrder.filter(
      (colId) => !allLocalColumnsParentIds.includes(colId)
    ),
    [localColumnOrder, allLocalColumnsParentIds]
  );
  const hasVisibilityChanged = (0, import_react.useMemo)(
    () => visibilitySettingsEnabled && (0, import_column_settings_utils.getHasVisibilityChanged)(columnVisibilityToReset, localColumnVisibility),
    [visibilitySettingsEnabled, columnVisibilityToReset, localColumnVisibility]
  );
  const hasColumnOrderChanged = columnOrderSettingsEnabled && (0, import_column_settings_utils.getHasColumnOrderChanged)(columnOrderToReset, filteredLocalColumnOrder);
  const hasColumnPinningChanged = columnPinningSettingsEnabled && (0, import_column_settings_utils.getHasColumnPinningChanged)(columnPinningToReset, localColumnPinning);
  const showHideAllState = (0, import_column_settings_utils.getCombinedVisibilityCheckboxState)(
    localColumnVisibility,
    leafColumnsWithoutBuiltin
  );
  const updateLocalColumnVisibility = (column, currentVisibilityState) => {
    const childColumns = column.columns;
    if (childColumns.length === 0) {
      setLocalColumnVisibility((prevState) => ({
        ...prevState,
        [column.id]: !currentVisibilityState
      }));
    } else {
      setLocalColumnVisibility(
        (prevState) => (0, import_column_settings_utils.getMultipleLocalColumnVisibility)(
          childColumns,
          currentVisibilityState,
          prevState
        )
      );
    }
    showResetStatus && setShowResetStatus(false);
  };
  const updateLocalColumnOrder = (newColumnOrder) => {
    setLocalColumnOrder(newColumnOrder);
    showResetStatus && setShowResetStatus(false);
  };
  const updateLocalColumnPinning = (newState) => {
    setLocalColumnPinning(newState);
    showResetStatus && setShowResetStatus(false);
  };
  const onShowHideAllChange = () => {
    setLocalColumnVisibility(
      (prevState) => (0, import_column_settings_utils.getMultipleLocalColumnVisibility)(
        leafColumnsWithoutBuiltin,
        showHideAllState,
        prevState
      )
    );
    showResetStatus && setShowResetStatus(false);
  };
  const showVisibilityControl = visibilitySettingsEnabled && activeFilterValue.length === 0;
  const showFilterControl = visibilitySettingsEnabled || columnPinningSettingsEnabled;
  const onReset = () => {
    let didReset = false;
    if (visibilitySettingsEnabled && columnVisibilityToReset) {
      setLocalColumnVisibility(
        (0, import_column_settings_utils.generateColumnVisibilityObject)(
          leafColumnsWithoutBuiltin,
          columnVisibilityToReset
        )
      );
      didReset = true;
    }
    if (columnOrderSettingsEnabled && columnOrderToReset) {
      const { columnOrderWithParents } = (0, import_column_settings_utils.generateColumnOrderWithParentIds)(
        table.getColumn,
        columnOrderToReset
      );
      setLocalColumnOrder(columnOrderWithParents);
      didReset = true;
    }
    if (columnPinningSettingsEnabled && columnPinningToReset) {
      setLocalColumnPinning(columnPinningToReset);
      didReset = true;
    }
    didReset && setShowResetStatus(true);
  };
  const onApply = () => {
    table.setColumnVisibility(localColumnVisibility);
    table.setColumnOrder(filteredLocalColumnOrder);
    table.setColumnPinning(localColumnPinning);
    onModalClose();
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_DataTableSettingsModal.DataTableSettingsModal,
    {
      title: intl.formatMessage(import_messages.dataTableMessages.columnSettingsTitle),
      show: true,
      showResetStatus,
      onDismiss: onModalClose,
      onReset,
      onApply,
      disableButtons: {
        reset: !hasColumnOrderChanged && !hasVisibilityChanged && !hasColumnPinningChanged
      },
      controlLabels: {
        reset: intl.formatMessage(
          import_messages.dataTableMessages.resetColumnSettingsButtonText
        ),
        cancel: intl.formatMessage(import_messages.dataTableMessages.cancelButtonText),
        apply: intl.formatMessage(import_messages.dataTableMessages.applyButtonText)
      },
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_DataTableVisibilityFilterControls.DataTableVisibilityFilterControls,
          {
            showVisibilityControl,
            showFilterControl,
            setActiveFilterValue,
            showHideAllState,
            onShowHideAllChange
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_DataTableColumnsList.DataTableColumnsList,
          {
            visibilitySettings: visibilitySettingsEnabled,
            columnOrderSettings: columnOrderSettingsEnabled,
            columnPinningSettings: columnPinningSettingsEnabled,
            activeFilterValue,
            localColumnVisibility,
            updateLocalColumnVisibility,
            localColumnOrder,
            updateLocalColumnOrder,
            localColumnPinning,
            updateLocalColumnPinning,
            hasFilterBar: showFilterControl || showVisibilityControl
          }
        )
      ]
    }
  );
};
function DataTableColumnSettingsModalFwd(props, forwardedRef) {
  const [modalOpen, setModalOpen] = (0, import_react.useState)(false);
  const [columnSettingsOverride, setColumnSettingsOverride] = (0, import_react.useState)();
  (0, import_react.useImperativeHandle)(forwardedRef, () => ({
    openColumnSettings: (columnSettings) => {
      setColumnSettingsOverride(columnSettings);
      setModalOpen(true);
    }
  }));
  if (modalOpen) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      ColumnSettingsModal,
      {
        ...props,
        columnSettingsOverride,
        onModalClose: () => setModalOpen(false)
      }
    );
  }
  return null;
}
const DataTableColumnSettingsModal = (0, import_react.forwardRef)(
  DataTableColumnSettingsModalFwd
);
