// @dynatrace/strato-components/tables/DataTable/components/Toolbar/ColumnSettings/DataTableColumnSettings.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DataTableColumnSettings_exports = {};
__export(DataTableColumnSettings_exports, {
  DataTableColumnSettings: () => DataTableColumnSettings
});
module.exports = __toCommonJS(DataTableColumnSettings_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_Button = require("../../../../../buttons/button/Button.js");
var import_Tooltip = require("../../../../../overlays/tooltip/Tooltip.js");
var import_DataTableContext = require("../../../contexts/DataTableContext.js");
var import_messages = require("../../../messages.js");
const DataTableColumnSettings = (props) => {
  const { customTriggerProps } = props;
  const {
    children: customTrigger,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    ...remainingProps
  } = customTriggerProps ?? {};
  const intl = (0, import_react_intl.useIntl)();
  const table = (0, import_DataTableContext.useDataTable)();
  const columnSettingsModal = (0, import_DataTableContext.useDataTableColumnSettings)();
  const columnSettingsButton = (0, import_react.useMemo)(() => {
    const hiddenColumnCount = Object.values(
      table.getState().columnVisibility || {}
    ).filter((value) => !value).length;
    const message = hiddenColumnCount === 0 ? intl.formatMessage(import_messages.dataTableMessages.columnSettingsButtonDefault) : intl.formatMessage(import_messages.dataTableMessages.columnSettingsButtonVisibility, {
      hiddenColumnCount
    });
    return {
      text: message,
      hiddenColumnCount
    };
  }, [table.getState().columnVisibility]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Tooltip.Tooltip, { text: intl.formatMessage(import_messages.dataTableMessages.columnSettingsTooltip), children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_Button.Button,
    {
      onClick: columnSettingsModal?.openColumnSettings,
      variant: "default",
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      ...remainingProps,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ColumnsIcon, {}) }),
        customTrigger ? customTrigger({
          hiddenColumnCount: columnSettingsButton.hiddenColumnCount,
          totalColumnCount: table.getAllColumns().length,
          defaultLabel: columnSettingsButton.text
        }) : columnSettingsButton.text
      ]
    }
  ) });
};
