// @dynatrace/strato-components/tables/DataTable/components/Toolbar/ColumnSettings/DataTableColumnsList.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DataTableColumnsList_exports = {};
__export(DataTableColumnsList_exports, {
  DataTableColumnsList: () => DataTableColumnsList
});
module.exports = __toCommonJS(DataTableColumnsList_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_virtual = require("@tanstack/react-virtual");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_column_settings_utils = require("./column-settings-utils.js");
var import_column_order_settings_dnd_utils = require("./ColumnOrderSettings/column-order-settings-dnd-utils.js");
var import_ColumnOrderSettingsDragAndDrop = require("./ColumnOrderSettings/ColumnOrderSettingsDragAndDrop.js");
var styles = __toESM(require("./DataTableColumnsList.sty.js"));
var import_DataTableColumnsListItem_css = require("./DataTableColumnsListItem.sty.js");
var import_DataTableColumnsListItem = require("./DataTableColumnsListItem.js");
var import_Grid = require("../../../../../layouts/grid/Grid.js");
var import_Text = require("../../../../../typography/text/Text.js");
var import_DataTableContext = require("../../../contexts/DataTableContext.js");
var import_ColumnVisibility = require("../../../features/ColumnVisibility/ColumnVisibility.js");
var import_messages = require("../../../messages.js");
const DataTableColumnsList = (props) => {
  const {
    visibilitySettings: showVisibilityCheckbox,
    columnOrderSettings,
    columnPinningSettings,
    activeFilterValue,
    localColumnVisibility,
    updateLocalColumnVisibility,
    localColumnOrder,
    updateLocalColumnOrder,
    localColumnPinning,
    updateLocalColumnPinning,
    hasFilterBar
  } = props;
  const table = (0, import_DataTableContext.useDataTable)();
  const intl = (0, import_react_intl.useIntl)();
  const parentRef = (0, import_react.useRef)(null);
  const allLeafColumnsWithoutBuiltin = table.getAllLeafColumnsWithoutBuiltin();
  const columnsById = (0, import_react.useMemo)(
    () => new Map(allLeafColumnsWithoutBuiltin.map((col) => [col.id, col])),
    [allLeafColumnsWithoutBuiltin]
  );
  const columnLabelsById = (0, import_react.useMemo)(
    () => (0, import_column_settings_utils.generateColumnLabelsMap)(allLeafColumnsWithoutBuiltin),
    [allLeafColumnsWithoutBuiltin]
  );
  const activeFilterWords = (0, import_react.useMemo)(
    () => (0, import_column_settings_utils.processFilterValue)(activeFilterValue),
    [activeFilterValue]
  );
  const compareColumns = (0, import_react.useCallback)(
    (a, b) => {
      const idChildColumnA = a.columns.at(0)?.id;
      const idChildColumnB = b.columns.at(0)?.id;
      const isBChildOfA = !!a.columns.find((childCol) => childCol.id === b.id);
      const isAChildOfB = !!b.columns.find((childCol) => childCol.id === a.id);
      if (isBChildOfA) {
        return -1;
      } else if (isAChildOfB) {
        return 1;
      }
      return localColumnOrder.indexOf(idChildColumnA ?? a.id) - localColumnOrder.indexOf(idChildColumnB ?? b.id);
    },
    [localColumnOrder]
  );
  const filterColumns = (0, import_react.useCallback)(
    (column, _, columnSection) => {
      if (activeFilterWords.length === 0) {
        return true;
      }
      const matchesFilter = (columnId) => {
        const columnLabel = columnLabelsById.get(columnId) ?? "";
        return activeFilterWords.every((word) => columnLabel.includes(word));
      };
      const childColumns = column.columns;
      if (childColumns.length === 0) {
        return matchesFilter(column.id);
      }
      return childColumns.filter((child) => columnSection.find((col) => col.id === child.id)).some((childColumn) => matchesFilter(childColumn.id));
    },
    [activeFilterWords, columnLabelsById]
  );
  const mapToListItem = (col, idx, pinnedPosition) => ({
    ...col,
    key: `${col.id}_${idx}_${pinnedPosition}`,
    // unique React key; parent columns may appear multiple times in list
    pinned: pinnedPosition
  });
  const orderedLeftFlatColumns = (0, import_react.useMemo)(() => {
    const localColumnsPinned = localColumnPinning.left?.map((id) => columnsById.get(id)).filter((column) => !!column) ?? [];
    const groupedColumns = (0, import_column_settings_utils.groupColumns)(localColumnsPinned);
    return groupedColumns.map((col, idx) => mapToListItem(col, idx, "left"));
  }, [columnsById, localColumnPinning.left]);
  const orderedCenterFlatColumns = (0, import_react.useMemo)(() => {
    const localColumnsUnpinned = allLeafColumnsWithoutBuiltin.filter(
      (column) => !localColumnPinning.left?.includes(column.id) && !localColumnPinning.right?.includes(column.id)
    );
    const groupedColumns = (0, import_column_settings_utils.groupColumns)(localColumnsUnpinned);
    return groupedColumns.toSorted(compareColumns).map((col, idx) => mapToListItem(col, idx, false));
  }, [
    allLeafColumnsWithoutBuiltin,
    compareColumns,
    localColumnPinning.left,
    localColumnPinning.right
  ]);
  const orderedRightFlatColumns = (0, import_react.useMemo)(() => {
    const localColumnsPinned = localColumnPinning.right?.map((id) => columnsById.get(id)).filter((column) => !!column) ?? [];
    const groupedColumns = (0, import_column_settings_utils.groupColumns)(localColumnsPinned);
    return groupedColumns.map((col, idx) => mapToListItem(col, idx, "right"));
  }, [columnsById, localColumnPinning.right]);
  const orderedFlatColumns = {
    left: orderedLeftFlatColumns,
    center: orderedCenterFlatColumns,
    right: orderedRightFlatColumns
  };
  const filteredColumns = (0, import_react.useMemo)(
    () => [
      ...orderedLeftFlatColumns.filter(filterColumns),
      ...orderedCenterFlatColumns.filter(filterColumns),
      ...orderedRightFlatColumns.filter(filterColumns)
    ],
    [
      filterColumns,
      orderedCenterFlatColumns,
      orderedLeftFlatColumns,
      orderedRightFlatColumns
    ]
  );
  const getItemKey = (0, import_react.useCallback)(
    (index) => filteredColumns[index].key,
    [filteredColumns]
  );
  const columnVirtualizer = (0, import_react_virtual.useVirtualizer)({
    count: filteredColumns.length,
    overscan: 5,
    getScrollElement: () => parentRef.current,
    getItemKey,
    // Approximate height of a column list item
    estimateSize: () => import_column_order_settings_dnd_utils.DATA_TABLE_COLUMNS_LIST_ITEM_MIN_HEIGHT,
    measureElement: (element) => {
      if (!element || !(element instanceof HTMLElement)) {
        return import_column_order_settings_dnd_utils.DATA_TABLE_COLUMNS_LIST_ITEM_MIN_HEIGHT;
      }
      const labelElement = element.querySelector(`.${import_DataTableColumnsListItem_css.columnName}`);
      if (!labelElement) {
        return import_column_order_settings_dnd_utils.DATA_TABLE_COLUMNS_LIST_ITEM_MIN_HEIGHT;
      }
      const computedStyle = getComputedStyle(element);
      const borderTop = parseFloat(computedStyle.borderTopWidth);
      const borderBottom = parseFloat(computedStyle.borderBottomWidth);
      return labelElement.getBoundingClientRect().height + 2 * 4 + // padding top and bottom
      borderTop + borderBottom;
    }
  });
  const virtualItems = columnVirtualizer.getVirtualItems();
  if (filteredColumns.length === 0) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { className: styles.noItemsFound, children: intl.formatMessage(import_messages.dataTableMessages.noItemsFound) });
  }
  const showColumnPinningButton = !!table.options.enableColumnPinning && columnPinningSettings;
  const canOrder = !!table.options.enableColumnOrdering && columnOrderSettings && activeFilterValue.length === 0;
  const columnListItems = virtualItems.map((virtualItem) => {
    const column = filteredColumns[virtualItem.index];
    const columnKey = column.key;
    const canHide = (0, import_ColumnVisibility.getCanHide)(table, column, import_column_settings_utils.isParentColumnHideable);
    const visibilityCheckboxState = (0, import_column_settings_utils.getVisibilityCheckboxState)(
      localColumnVisibility,
      column
    );
    const showColumnOrderDragHandle = canOrder && column.pinned === false;
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_DataTableColumnsListItem.DataTableColumnsListItem,
      {
        column,
        columnIndex: localColumnOrder.indexOf(column.id),
        orderedFlatColumns,
        showVisibilityCheckbox,
        enableVisibilityCheckbox: canHide,
        currentVisibilityState: visibilityCheckboxState,
        onVisibilityChange: updateLocalColumnVisibility,
        showColumnOrderDragHandle,
        showColumnPinningMenu: showColumnPinningButton,
        localColumnPinning,
        updateLocalColumnPinning,
        virtualIndex: virtualItem.index,
        measureElement: columnVirtualizer.measureElement
      },
      columnKey
    );
  });
  const columnList = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      ref: parentRef,
      className: (0, import_clsx.default)(
        "strato-table-settings-virtualization-container",
        styles.virtualizationScrollContainer({
          hasFilter: hasFilterBar
        })
      ),
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_Grid.Grid,
        {
          gridTemplateColumns: "1fr",
          padding: 0,
          gap: 0,
          className: (0, import_clsx.default)(styles.virtualizationInnerContainer),
          gridTemplateRows: columnVirtualizer.measurementsCache.map((el) => `${el.size}px`).join(" "),
          height: columnVirtualizer.getTotalSize() + 1,
          children: columnListItems
        }
      )
    }
  );
  return canOrder ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ColumnOrderSettingsDragAndDrop.ColumnOrderSettingsDragAndDropContext,
    {
      localColumnOrder,
      updateLocalColumnOrder,
      localColumnPinning,
      itemVirtualizer: columnVirtualizer,
      children: columnList
    }
  ) : columnList;
};
