// @dynatrace/strato-geo/map/components/toolbar/buttons/MapDownloadDataButton.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MapDownloadDataButton_exports = {};
__export(MapDownloadDataButton_exports, {
  MapDownloadDataButton: () => MapDownloadDataButton
});
module.exports = __toCommonJS(MapDownloadDataButton_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_charts = require("@dynatrace/strato-components/charts");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_use_map_raw_data = require("../../../hooks/use-map-raw-data.js");
var import_build_and_download_csv = require("../../../utils/build-and-download-csv.js");
var import_constants = require("../constants.js");
var import_use_a11y_toolbar_tab_index = require("../hooks/use-a11y-toolbar-tab-index.js");
const MapDownloadDataButton = () => {
  const downloadDataButtonRef = (0, import_react.useRef)(null);
  const intl = (0, import_react_intl.useIntl)();
  const tabIndex = (0, import_use_a11y_toolbar_tab_index.useA11yToolbarTabIndex)("download-data");
  const data = (0, import_use_map_raw_data.useMapRawData)();
  const handleDownloadDataKeyboard = (event) => {
    if (event.key === "Enter") {
      (0, import_build_and_download_csv.buildAndDownloadCsv)(data);
    }
  };
  const handleDownloadDataToolbar = () => {
    (0, import_build_and_download_csv.buildAndDownloadCsv)(data);
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_charts._ChartToolbarRenderer.Control,
    {
      ref: downloadDataButtonRef,
      prefixIcon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.DownloadIcon, {}),
      tooltipText: intl.formatMessage(import_constants.TOOLBAR_MESSAGES.downloadData),
      testId: "map-toolbar-download",
      onKeyDown: handleDownloadDataKeyboard,
      onClick: handleDownloadDataToolbar,
      tabIndex,
      placement: "toolbar"
    }
  );
};
