// @dynatrace/strato-geo/map/components/toolbar/buttons/MapResetButton.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MapResetButton_exports = {};
__export(MapResetButton_exports, {
  MapResetButton: () => MapResetButton
});
module.exports = __toCommonJS(MapResetButton_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_maplibre = require("@vis.gl/react-maplibre");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_charts = require("@dynatrace/strato-components/charts");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_use_initial_view_context = require("../../../hooks/use-initial-view-context.js");
var import_store = require("../../../store/store.js");
var import_constants = require("../constants.js");
var import_use_a11y_toolbar_tab_index = require("../hooks/use-a11y-toolbar-tab-index.js");
const MapResetButton = () => {
  const { current: map } = (0, import_react_maplibre.useMap)();
  const { longitude = 0, latitude = 0, zoom = 0 } = (0, import_use_initial_view_context.useInitialViewContext)();
  const intl = (0, import_react_intl.useIntl)();
  const { resetEnabled } = (0, import_store.useMapZoomState)();
  const setZoomState = (0, import_store.useSetStateZoom)();
  const resetButtonRef = (0, import_react.useRef)(null);
  const tabIndex = (0, import_use_a11y_toolbar_tab_index.useA11yToolbarTabIndex)("reset");
  const handleReset = () => {
    map?.flyTo({ center: [longitude, latitude], zoom });
    setZoomState({ zoomToFitEnabled: true, resetEnabled: false });
  };
  const handleKeyDown = (event) => {
    if (event.code === "Enter") {
      handleReset();
    }
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_charts._ChartToolbarRenderer.Control,
    {
      ref: resetButtonRef,
      prefixIcon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ResetIcon, {}),
      tooltipText: intl.formatMessage(import_constants.TOOLBAR_MESSAGES.zoomActionReset),
      testId: "map-toolbar-reset",
      onKeyDown: handleKeyDown,
      onClick: handleReset,
      tabIndex,
      disabled: !resetEnabled
    }
  );
};
