// @dynatrace/strato-geo/map/components/toolbar/buttons/MapZoomInOutButtons.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MapZoomInOutButtons_exports = {};
__export(MapZoomInOutButtons_exports, {
  MapZoomInOutButtons: () => MapZoomInOutButtons
});
module.exports = __toCommonJS(MapZoomInOutButtons_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_maplibre = require("@vis.gl/react-maplibre");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_charts = require("@dynatrace/strato-components/charts");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_selectors = require("../../../store/selectors.js");
var import_constants = require("../constants.js");
var import_use_a11y_toolbar_tab_index = require("../hooks/use-a11y-toolbar-tab-index.js");
const MapZoomInOutButtons = () => {
  const { current: map } = (0, import_react_maplibre.useMap)();
  const { zoomInEnabled, zoomOutEnabled } = (0, import_selectors.useMapZoomState)();
  const zoomInButtonRef = (0, import_react.useRef)(null);
  const zoomButtonOutRef = (0, import_react.useRef)(null);
  const intl = (0, import_react_intl.useIntl)();
  const tabIndexIn = (0, import_use_a11y_toolbar_tab_index.useA11yToolbarTabIndex)("zoom-in");
  const tabIndexOut = (0, import_use_a11y_toolbar_tab_index.useA11yToolbarTabIndex)("zoom-out");
  const handleZoomKeyboardMode = (event, zoomBehavior) => {
    if (event.code === "Enter") {
      handleZoomStepToolbarMode(zoomBehavior);
    }
  };
  const handleZoomStepToolbarMode = (zoomBehavior) => {
    if (zoomBehavior === "zoom-in") {
      map?.zoomIn();
    }
    if (zoomBehavior === "zoom-out") {
      map?.zoomOut();
    }
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_charts._ChartToolbarRenderer.Control,
      {
        ref: zoomInButtonRef,
        prefixIcon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ZoomInIcon, {}),
        tooltipText: intl.formatMessage(import_constants.TOOLBAR_MESSAGES.zoomActionZoomIn),
        testId: "map-toolbar-zoom-in",
        onKeyDown: (event) => handleZoomKeyboardMode(event, "zoom-in"),
        onClick: () => handleZoomStepToolbarMode("zoom-in"),
        tabIndex: tabIndexIn,
        disabled: !zoomInEnabled
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_charts._ChartToolbarRenderer.Control,
      {
        ref: zoomButtonOutRef,
        prefixIcon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ZoomOutIcon, {}),
        tooltipText: intl.formatMessage(import_constants.TOOLBAR_MESSAGES.zoomActionZoomOut),
        testId: "map-toolbar-zoom-out",
        onKeyDown: (event) => handleZoomKeyboardMode(event, "zoom-out"),
        onClick: () => handleZoomStepToolbarMode("zoom-out"),
        tabIndex: tabIndexOut,
        disabled: !zoomOutEnabled
      }
    )
  ] });
};
