// @dynatrace/strato-geo/map/components/toolbar/buttons/MapZoomToFitButton.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MapZoomToFitButton_exports = {};
__export(MapZoomToFitButton_exports, {
  MapZoomToFitButton: () => MapZoomToFitButton
});
module.exports = __toCommonJS(MapZoomToFitButton_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_maplibre = require("@vis.gl/react-maplibre");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_charts = require("@dynatrace/strato-components/charts");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_constants = require("../../../constants.js");
var import_use_data_bounding_box = require("../../../hooks/use-data-bounding-box.js");
var import_store = require("../../../store/store.js");
var import_constants2 = require("../constants.js");
var import_use_a11y_toolbar_tab_index = require("../hooks/use-a11y-toolbar-tab-index.js");
const MapZoomToFitButton = () => {
  const resetButtonRef = (0, import_react.useRef)(null);
  const { current: map } = (0, import_react_maplibre.useMap)();
  const tabIndex = (0, import_use_a11y_toolbar_tab_index.useA11yToolbarTabIndex)("zoom-to-fit");
  const intl = (0, import_react_intl.useIntl)();
  const boundingBox = (0, import_use_data_bounding_box.useDataBoundingBox)();
  const { zoomToFitEnabled } = (0, import_store.useMapZoomState)();
  const setZoomState = (0, import_store.useSetStateZoom)();
  const handleZoomToFit = () => {
    map?.fitBounds(boundingBox, import_constants.FIT_BOUNDS_OPTIONS);
    setZoomState({ zoomToFitEnabled: false, zoomToFitInProgress: true });
  };
  const handleKeyDown = (event) => {
    if (event.code === "Enter") {
      handleZoomToFit();
    }
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_charts._ChartToolbarRenderer.Control,
    {
      ref: resetButtonRef,
      prefixIcon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ZoomToFitIcon, {}),
      tooltipText: intl.formatMessage(import_constants2.TOOLBAR_MESSAGES.zoomActionZoomToFit),
      testId: "map-toolbar-zoom-to-fit",
      onKeyDown: handleKeyDown,
      onClick: handleZoomToFit,
      tabIndex,
      disabled: !zoomToFitEnabled
    }
  );
};
MapZoomToFitButton["displayName"] = "MapZoomToFit";
