// @dynatrace/strato-components/esm/charts/gauge/components/GaugeChartLayout/GaugeChartToolbar.js
import { jsx } from "react/jsx-runtime";
import { useTooltipReducer } from "../../../core/components/overlay-tooltip/useTooltipReducer.js";
import { ChartToolbarRenderer } from "../../../core/components/toolbar/ChartToolbarRenderer.js";
import { getA11yToolbarConfig } from "../../../core/utils/zoom.js";
import { zIndexGaugeAreas } from "../../constants.js";
const GaugeChartToolbar = (props) => {
  const dispatch = useTooltipReducer();
  const { intents, toolbar } = props;
  if (toolbar.hidden) {
    return null;
  }
  return /* @__PURE__ */ jsx(
    ChartToolbarRenderer,
    {
      visible: false,
      collapsed: false,
      draggable: true,
      placement: toolbar.placement,
      initialA11yState: getA11yToolbarConfig({
        intent: !!intents
      }),
      zIndex: zIndexGaugeAreas.toolbar,
      intents,
      menuDisplayMode: "never",
      toolbarDisplayMode: "overlay",
      removeDragHandle: true,
      visibleWhenMouseInBounds: true,
      onToggleMenu: (isOpen) => {
        if (isOpen) {
          dispatch({ type: "RESET_TOOLTIP" })();
        }
      }
    }
  );
};
export {
  GaugeChartToolbar
};
