// @dynatrace/strato-components/esm/charts/gauge/components/GaugeChartLayout/GaugeChartLayout.js
import { jsx, jsxs } from "react/jsx-runtime";
import { isEmpty } from "lodash-es";
import { forwardRef } from "react";
import { layoutContainer } from "./GaugeChartLayout.sty.js";
import { accessValue } from "../../../../core/utils/accessors/access-value.js";
import { DEFAULT_VALUE_ACCESSOR } from "../../../core/constants.js";
import { useGaugeChartColorContext } from "../../hooks/useGaugeChartColorContext.js";
import { useGaugeChartLoadingContext } from "../../hooks/useGaugeChartLoadingContext.js";
import { useGaugeChartRangeMaxContext } from "../../hooks/useGaugeChartRangeMaxContext.js";
import { useGaugeChartRangeMinContext } from "../../hooks/useGaugeChartRangeMinContext.js";
import { useGaugeChartThresholdContext } from "../../hooks/useGaugeChartThresholdContext.js";
import { GaugeBaseArc } from "../GaugeBaseArc.js";
import { GaugeLabels } from "../GaugeLabels.js";
import { GaugeChartOverlay } from "../GaugeLayoutOverlay/GaugeLayoutOverlay.js";
import { GaugeRootSVG } from "../GaugeRootSVG.js";
import { GaugeThresholdIndicators } from "../GaugeThresholdIndicators.js";
import { GaugeValueArc } from "../GaugeValueArc.js";
const GaugeChartLayout = forwardRef(({ value, prefixIcon, intents, toolbar, valueAccessor }, forwardedRef) => {
  const thresholds = useGaugeChartThresholdContext();
  const min = useGaugeChartRangeMinContext();
  const max = useGaugeChartRangeMaxContext();
  const valueArcColor = useGaugeChartColorContext();
  const showLoading = useGaugeChartLoadingContext();
  const displayValue = typeof value === "number" ? value : accessValue(value, valueAccessor ?? DEFAULT_VALUE_ACCESSOR) ?? 0;
  const toolbarConfig = {
    ...toolbar,
    hidden: toolbar?.hidden || showLoading
  };
  return /* @__PURE__ */ jsx(
    "div",
    {
      "data-testid": "gauge-layout-container",
      ref: forwardedRef,
      className: layoutContainer,
      children: /* @__PURE__ */ jsx(
        GaugeChartOverlay,
        {
          value: displayValue,
          color: valueArcColor,
          intents,
          showLoading,
          toolbar: toolbarConfig,
          children: /* @__PURE__ */ jsxs(GaugeRootSVG, { children: [
            /* @__PURE__ */ jsx(GaugeBaseArc, {}),
            /* @__PURE__ */ jsx(
              GaugeValueArc,
              {
                value: displayValue,
                color: valueArcColor,
                min,
                max
              }
            ),
            !isEmpty(thresholds) && /* @__PURE__ */ jsx(GaugeThresholdIndicators, {}),
            /* @__PURE__ */ jsx(GaugeLabels, { value: displayValue, prefixIcon })
          ] })
        }
      )
    }
  );
});
GaugeChartLayout.displayName = "GaugeChartLayout";
export {
  GaugeChartLayout
};
