// @dynatrace/strato-components/tables/SimpleTable/SimpleTable.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SimpleTable_exports = {};
__export(SimpleTable_exports, {
  SimpleTable: () => SimpleTable
});
module.exports = __toCommonJS(SimpleTable_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_table = require("@tanstack/react-table");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_useColumns = require("./hooks/useColumns.js");
var styles = __toESM(require("./SimpleTable.sty.js"));
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useSafeForwardProps = require("../../core/hooks/useSafeForwardProps.js");
const COMPONENT_NAME = "SimpleTable";
function InternalSimpleTable(props, forwardedRef) {
  const {
    data,
    columns,
    variant,
    style: consumerStyle,
    className: consumerClassName,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const ariaLabelingProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(restProps);
  const forwardProps = (0, import_useSafeForwardProps.useSafeForwardProps)(restProps, ariaLabelingProps);
  const extractedColumns = (0, import_useColumns.useColumns)(columns);
  const table = (0, import_react_table.useReactTable)({
    data,
    columns: extractedColumns,
    getCoreRowModel: (0, import_react_table.getCoreRowModel)()
  });
  const combinedVariants = {
    rowSeparation: variant?.rowSeparation ?? "horizontalDividers",
    verticalDividers: variant?.verticalDividers ?? false,
    contained: variant?.contained ?? true,
    rowDensity: variant?.rowDensity ?? "default",
    fontStyle: variant?.fontStyle ?? "text"
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    "table",
    {
      ref: forwardedRef,
      className: (0, import_clsx.default)(styles.simpleTable, consumerClassName, {
        "strato-table-contained": combinedVariants.contained,
        "strato-table-colsep-divider": combinedVariants.verticalDividers,
        "strato-table-code-font": combinedVariants.fontStyle === "code",
        "strato-table-cell-density-default": combinedVariants.rowDensity === "default",
        "strato-table-cell-density-condensed": combinedVariants.rowDensity === "condensed",
        "strato-table-cell-density-comfortable": combinedVariants.rowDensity === "comfortable",
        "strato-table-rowsep-zebra": combinedVariants.rowSeparation === "zebraStripes",
        "strato-table-rowsep-divider": combinedVariants.rowSeparation === "horizontalDividers" || combinedVariants.rowSeparation === "zebraStripes"
      }),
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...ariaLabelingProps,
      ...behavioralTrackingProps,
      ...forwardProps,
      "data-dt-component": COMPONENT_NAME,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)("thead", { children: table.getHeaderGroups().map((headerGroup) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)("tr", { className: "strato-table-header", children: headerGroup.headers.map((header) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          "th",
          {
            className: (0, import_clsx.default)("strato-table-header-cell", {
              "strato-table-cell-alignment-left": header.column.columnDef.meta?.alignment === "left",
              "strato-table-cell-alignment-center": header.column.columnDef.meta?.alignment === "center",
              "strato-table-cell-alignment-right": header.column.columnDef.meta?.alignment === "right",
              "strato-table-cell-alignment-top-left": header.column.columnDef.meta?.alignment === "top-left",
              "strato-table-cell-alignment-top-center": header.column.columnDef.meta?.alignment === "top-center",
              "strato-table-cell-alignment-top-right": header.column.columnDef.meta?.alignment === "top-right",
              "strato-table-cell-alignment-bottom-left": header.column.columnDef.meta?.alignment === "bottom-left",
              "strato-table-cell-alignment-bottom-center": header.column.columnDef.meta?.alignment === "bottom-center",
              "strato-table-cell-alignment-bottom-right": header.column.columnDef.meta?.alignment === "bottom-right"
            }),
            style: {
              minWidth: header.column.columnDef.meta?.minWidth,
              width: header.column.columnDef.meta?.width,
              maxWidth: header.column.columnDef.meta?.maxWidth
            },
            children: header.isPlaceholder ? null : (0, import_react_table.flexRender)(
              header.column.columnDef.header,
              header.getContext()
            )
          },
          header.id
        )) }, headerGroup.id)) }),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)("tbody", { children: table.getRowModel().rows.map((row, rowIndex) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          "tr",
          {
            className: (0, import_clsx.default)("strato-table-row", {
              "strato-table-row-even": rowIndex % 2
            }),
            children: row.getVisibleCells().map((cell) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              "td",
              {
                className: (0, import_clsx.default)("strato-table-cell", {
                  "strato-table-cell-alignment-left": cell.column.columnDef.meta?.alignment === "left",
                  "strato-table-cell-alignment-center": cell.column.columnDef.meta?.alignment === "center",
                  "strato-table-cell-alignment-right": cell.column.columnDef.meta?.alignment === "right",
                  "strato-table-cell-alignment-top-left": cell.column.columnDef.meta?.alignment === "top-left",
                  "strato-table-cell-alignment-top-center": cell.column.columnDef.meta?.alignment === "top-center",
                  "strato-table-cell-alignment-top-right": cell.column.columnDef.meta?.alignment === "top-right",
                  "strato-table-cell-alignment-bottom-left": cell.column.columnDef.meta?.alignment === "bottom-left",
                  "strato-table-cell-alignment-bottom-center": cell.column.columnDef.meta?.alignment === "bottom-center",
                  "strato-table-cell-alignment-bottom-right": cell.column.columnDef.meta?.alignment === "bottom-right"
                }),
                style: {
                  minWidth: cell.column.columnDef.meta?.minWidth,
                  width: cell.column.columnDef.meta?.width,
                  maxWidth: cell.column.columnDef.meta?.maxWidth
                },
                children: (0, import_react_table.flexRender)(cell.column.columnDef.cell, cell.getContext())
              },
              cell.id
            ))
          },
          row.id
        )) })
      ]
    }
  );
}
InternalSimpleTable.displayName = COMPONENT_NAME;
const SimpleTable = /* @__PURE__ */ (0, import_react.forwardRef)(InternalSimpleTable);
