// @dynatrace/strato-components/esm/charts/xy-chart/components/tooltip/HoveredPointTooltip/hoveredPointTooltipHeaderUtils.js
import { getDefaultAction } from "../../../../core/components/series-action/hooks/useDefaultAction.js";
import { getFormatter } from "../../../../core/utils/get-formatter.js";
import { safeGet } from "../../../../core/utils/safe-get.js";
import { buildBarSeriesActionItem } from "../../../utils/build-series-action-item.js";
import { getColorFromScaleAndData } from "../../../utils/coloring/get-color-from-scale-and-data.js";
import {
  isBarTooltipContext,
  isRectTooltipContext
} from "../../../utils/type-guards.js";
const buildHoveredPointTooltipHeaderProps = ({
  datapoint,
  series,
  seriesToMetadata,
  seriesWithDatapoints,
  datapointMetadataMap,
  colorScaleMap,
  truncationMode,
  idFactory = () => crypto.randomUUID()
}) => {
  let content;
  let valueToDisplay = "";
  let colorToDisplay = "";
  let actions;
  const metadata = safeGet(seriesToMetadata, series);
  const enrichedDatapoint = {
    series,
    datapoint,
    metadata
  };
  if (isRectTooltipContext(enrichedDatapoint)) {
    const colorScale = safeGet(colorScaleMap, series.id);
    const datapointMetadata = safeGet(datapointMetadataMap, datapoint);
    const rawValue = enrichedDatapoint.datapoint.value;
    const { valueAccessorLabel, valueAccessor, valueFormatter, valueUnit } = enrichedDatapoint.metadata;
    const tooltipName = valueAccessorLabel ?? valueAccessor ?? "";
    content = {
      id: idFactory(),
      name: tooltipName
    };
    valueToDisplay = typeof rawValue === "number" ? getFormatter(valueFormatter)(rawValue, valueUnit) : rawValue?.toString() ?? "";
    colorToDisplay = rawValue !== void 0 ? getColorFromScaleAndData(
      colorScale,
      enrichedDatapoint.datapoint.value,
      datapointMetadata.colorOverride
    ) ?? "" : "";
    const metadataActions = enrichedDatapoint.metadata.actions;
    actions = metadataActions ? () => metadataActions(enrichedDatapoint.datapoint) : void 0;
  } else if (isBarTooltipContext(enrichedDatapoint)) {
    const { color, name } = metadata;
    content = {
      id: idFactory(),
      name: name ?? ""
    };
    colorToDisplay = color?.Default ?? "";
    const metadataActions = enrichedDatapoint.metadata.actions;
    actions = metadataActions ? () => {
      const actionItem = buildBarSeriesActionItem({
        datapoint: enrichedDatapoint.datapoint,
        allSeries: seriesWithDatapoints,
        seriesToMetadata,
        datapointMetadataMap,
        currentSeries: enrichedDatapoint.series
      });
      return metadataActions(actionItem);
    } : void 0;
  } else {
    content = {
      id: idFactory(),
      name: ""
    };
  }
  return {
    color: colorToDisplay,
    name: content.name,
    value: valueToDisplay,
    actions,
    defaultActions: getDefaultAction(content),
    truncationMode
  };
};
export {
  buildHoveredPointTooltipHeaderProps
};
