// @dynatrace/strato-components/esm/charts/xy-chart/components/tooltip/HoveredPointTooltip/HoveredPointTooltip.js
import { jsx, jsxs } from "react/jsx-runtime";
import { useIntl } from "react-intl";
import { buildHoveredPointTooltipBodyProps } from "./hoveredPointTooltipBodyUtils.js";
import { buildHoveredPointTooltipHeaderProps } from "./hoveredPointTooltipHeaderUtils.js";
import { OverlayTooltip } from "../../../../core/components/overlay-tooltip/OverlayTooltip.js";
import { useAxesFormatter } from "../../../hooks/useAxesFormatter.js";
import { useBarData } from "../../../hooks/useBarData.js";
import { useFormatCategory } from "../../../hooks/useFormatCategory.js";
import { useXYChartColorScales } from "../../../hooks/useXYChartColorScales.js";
import { useXYChartData } from "../../../hooks/useXYChartData.js";
import { useXYChartTruncationMode } from "../../../hooks/useXYChartTruncationMode.js";
function HoveredPointTooltip({
  datapoint,
  series,
  seriesToMetadata
}) {
  const { filteredSeriesWithDatapoints } = useBarData();
  const { axisIdToScaleDomain, datapointMetadataMap } = useXYChartData();
  const colorScaleMap = useXYChartColorScales();
  const { truncationMode } = useXYChartTruncationMode();
  const axisIdToFormatter = useAxesFormatter();
  const formatCategory = useFormatCategory();
  const intl = useIntl();
  const { name, value, color, actions } = buildHoveredPointTooltipHeaderProps({
    datapoint,
    series,
    seriesToMetadata,
    seriesWithDatapoints: filteredSeriesWithDatapoints,
    datapointMetadataMap,
    colorScaleMap
  });
  const { yAxisValue, xAxisValue, xAxisDuration, yAxisDuration } = buildHoveredPointTooltipBodyProps({
    datapoint,
    series,
    seriesToMetadata,
    axisIdToFormatter,
    axisIdToScaleDomain,
    datapointMetadataMap,
    formatCategory,
    intl
  });
  const isRect = series.type === "rect";
  const yText = `${yAxisValue} ${yAxisDuration ? `(${yAxisDuration})` : ""}`;
  const data = [
    {
      name: isRect ? yText : name,
      value: isRect ? value : yText,
      color
    }
  ];
  const label = datapoint.x0 instanceof Date ? "" : void 0;
  return /* @__PURE__ */ jsxs(OverlayTooltip, { truncation: truncationMode, children: [
    /* @__PURE__ */ jsx(
      OverlayTooltip.Header,
      {
        header1: xAxisValue,
        duration1: xAxisDuration,
        label
      }
    ),
    /* @__PURE__ */ jsx(
      OverlayTooltip.Series,
      {
        data,
        variant: "shared",
        activeItem: -1,
        seriesItemTemplate: (item) => /* @__PURE__ */ jsx(OverlayTooltip.SeriesItem, { ...item, actions })
      }
    )
  ] });
}
export {
  HoveredPointTooltip
};
