// @dynatrace/strato-components/esm/charts/xy-chart/components/tooltip/HoveredPointTooltip/hoveredPointTooltipBodyUtils.js
import { safeGet } from "../../../../core/utils/safe-get.js";
import { isCategorical } from "../../../types/guards.js";
import { getRelativeAccessor } from "../../../utils/value-representation.js";
import { getTooltipFormatter } from "../utils/get-tooltip-formatter.js";
import { getHoveredPointTooltipBodyContent } from "./hooks/getHoveredPointTooltipBodyContent.js";
const buildHoveredPointTooltipBodyProps = ({
  datapoint,
  series,
  seriesToMetadata,
  axisIdToFormatter,
  axisIdToScaleDomain,
  datapointMetadataMap,
  formatCategory,
  intl
}) => {
  const { xAxis, yAxis } = safeGet(seriesToMetadata, series);
  const xFormatter = getTooltipFormatter(xAxis, axisIdToFormatter, intl);
  const yFormatter = getTooltipFormatter(yAxis, axisIdToFormatter, intl);
  const datapointMetadata = datapointMetadataMap.get(datapoint);
  const binDomain = datapointMetadata?.binDomain;
  const isRelative = typeof binDomain === "number" && binDomain;
  const yDomain = isRelative && datapointMetadata?.baseline === "y" ? {
    min: binDomain * -1,
    max: binDomain
  } : safeGet(axisIdToScaleDomain, xAxis.id);
  const xDomain = isRelative && datapointMetadata?.baseline === "x" ? {
    min: binDomain * -1,
    max: binDomain
  } : safeGet(axisIdToScaleDomain, yAxis.id);
  const { x0, x1, y0, y1 } = datapoint;
  const buildCategoricalStringFormatter = (axis) => isCategorical(axis.type) ? (value) => formatCategory(axis, value) ?? value : void 0;
  const xResolvedFormatter = buildCategoricalStringFormatter(xAxis) ?? xFormatter;
  const yResolvedFormatter = buildCategoricalStringFormatter(yAxis) ?? yFormatter;
  const { text: textX, duration: durationX } = getHoveredPointTooltipBodyContent(
    intl,
    x0,
    x1,
    xResolvedFormatter,
    getRelativeAccessor(xAxis, yDomain)
  );
  const { text: textY, duration: durationY } = getHoveredPointTooltipBodyContent(
    intl,
    y0,
    y1,
    yResolvedFormatter,
    getRelativeAccessor(yAxis, xDomain)
  );
  return {
    xAxisLabel: xAxis.label,
    yAxisLabel: yAxis.label,
    xAxisValue: textX,
    yAxisValue: textY,
    xAxisDuration: durationX,
    yAxisDuration: durationY
  };
};
export {
  buildHoveredPointTooltipBodyProps
};
