// @dynatrace/strato-components/editors/dql-editor/dql/dql-language.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var dql_language_exports = {};
__export(dql_language_exports, {
  getDqlConfig: () => getDqlConfig
});
module.exports = __toCommonJS(dql_language_exports);
var import_autocomplete = require("@codemirror/autocomplete");
var import_language = require("@codemirror/language");
var import_lint = require("@codemirror/lint");
var import_view = require("@codemirror/view");
var import_dql_linter = require("./dql-linter.js");
var import_dql = require("./dql.js");
var import_open_autocomplete = require("./open-autocomplete.js");
var import_suggestion_autocompletions = require("./suggestion-autocompletions.js");
const VALIDATION_DELAY = 500;
function getDqlConfig(readOnly, onValidationStart, onValidationEnd, setValidityInfo) {
  const language = import_language.StreamLanguage.define(import_dql.dql);
  return {
    language,
    extendWith: [
      language.data.of({
        commentTokens: {
          block: {
            open: "/*",
            close: "*/"
          },
          line: "//"
        },
        closeBrackets: {
          brackets: ["(", "[", "{", '"'],
          before: ")]}:;>|",
          stringPrefixes: []
        }
      }),
      import_view.EditorView.domEventObservers({
        // Clear lint decorations when composition starts. Lint spans in the DOM
        // corrupt CodeMirror's DOM reconciliation during IME / dead-key composition.
        // The linter re-runs automatically after composition ends since the
        // composed character is a document change.
        compositionstart: (_event, view) => {
          (0, import_autocomplete.closeCompletion)(view);
          view.dispatch((0, import_lint.setDiagnostics)(view.state, []));
        }
      }),
      (0, import_lint.linter)(
        async (view) => {
          const editorContent = view.state.doc.toString();
          const validityResult = await (0, import_dql_linter.dqlLinter)(
            editorContent,
            onValidationStart,
            onValidationEnd
          );
          if (view.composing) {
            return [];
          }
          setValidityInfo?.(validityResult);
          return validityResult;
        },
        {
          delay: VALIDATION_DELAY
        }
      ),
      import_view.EditorView.updateListener.of((update) => {
        const isTyping = update.docChanged && update.transactions.some(
          (transaction) => transaction.isUserEvent("input.type")
        );
        const isDeleting = update.docChanged && update.transactions.some(
          (transaction) => transaction.isUserEvent("delete.backward")
        );
        const isFocused = update.focusChanged && update.view.hasFocus;
        if (!(isTyping || isDeleting || isFocused)) {
          return;
        }
        if (update.view.composing) {
          return;
        }
        void (0, import_open_autocomplete.openAutocomplete)(
          {
            isTyping,
            isDeleting,
            isFocused
          },
          update.state
        ).then((shouldComplete) => {
          if (shouldComplete) {
            (0, import_autocomplete.startCompletion)(update.view);
          }
        });
      })
    ],
    autocompleteOptions: readOnly ? void 0 : {
      activateOnTyping: false,
      optionClass: (completion) => completion.label.length === 0 ? "cm-empty-option" : "",
      providers: [import_suggestion_autocompletions.suggestionAutocompletion]
    }
  };
}
