// @dynatrace/strato-components/editors/dql-editor/dql/dql-linter.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var dql_linter_exports = {};
__export(dql_linter_exports, {
  dqlLinter: () => dqlLinter,
  verifyQuery: () => verifyQuery
});
module.exports = __toCommonJS(dql_linter_exports);
var import_lodash_es = require("lodash");
var import_client_query = require("@dynatrace-sdk/client-query");
const severities = {
  error: "error",
  warning: "warning",
  info: "info"
};
const toSeverity = (severity) => severities[severity.toLowerCase()] || "info";
const toLinterMessage = ({ notifications }) => (notifications ?? []).map(
  (notification) => notification.syntaxPosition && notification.severity ? {
    from: notification.syntaxPosition.start.index,
    to: notification.syntaxPosition.end.index + 1,
    message: notification.message ?? "",
    severity: toSeverity(notification.severity)
  } : void 0
).filter((x) => x !== void 0);
const transformToValidityInfo = (diagnostics) => diagnostics.map(
  (diagnostic) => ({
    ...diagnostic,
    severity: diagnostic.severity === "hint" ? "info" : diagnostic.severity
  })
);
const rawVerify = async (query) => {
  return await import_client_query.queryAssistanceClient.queryVerify({ body: { query } });
};
const TTL_MS = 3 * 1e3;
const memoizedRawVerify = (0, import_lodash_es.memoize)(rawVerify);
const ttlTimers = /* @__PURE__ */ new Map();
const verifyQuery = async (query) => {
  try {
    if (!query.length) {
      return void 0;
    }
    const result = await memoizedRawVerify(query);
    const pendingTimeout = ttlTimers.get(query);
    if (pendingTimeout) {
      clearTimeout(pendingTimeout);
    }
    const timeout = setTimeout(() => {
      memoizedRawVerify.cache.delete(query);
      ttlTimers.delete(query);
    }, TTL_MS);
    ttlTimers.set(query, timeout);
    return result;
  } catch {
    memoizedRawVerify.cache.delete(query);
    const timeout = ttlTimers.get(query);
    if (timeout) {
      clearTimeout(timeout);
      ttlTimers.delete(query);
    }
  }
  return void 0;
};
const dqlLinter = async (content, onValidationStart, onValidationEnd) => {
  onValidationStart?.();
  const result = await verifyQuery(content);
  const dqlErrors = result ? toLinterMessage(result) : [];
  const dqlEditorValidityInfos = transformToValidityInfo(dqlErrors);
  onValidationEnd?.(dqlEditorValidityInfos);
  return dqlEditorValidityInfos;
};
