// @dynatrace/strato-components/editors/dql-editor/dql/SuggestionPart.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SuggestionPart_exports = {};
__export(SuggestionPart_exports, {
  SuggestionPart: () => SuggestionPart
});
module.exports = __toCommonJS(SuggestionPart_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_DocumentationLink = require("./DocumentationLink.js");
var import_SuggestionPart_css = require("./SuggestionPart.sty.js");
var import_CodeSnippet = require("../../../content/code-snippet/CodeSnippet.js");
var import_Flex = require("../../../layouts/flex/Flex.js");
var import_safe_sprinkles = require("../../../styles/safe-sprinkles.js");
var import_Text = require("../../../typography/text/Text.js");
const SuggestionPart = (props) => {
  const { suggestionPart } = props;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { flexDirection: "column", gap: 0, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Text.Text, { textStyle: "base", className: import_SuggestionPart_css.suggestionPartSectionInfoCSS, children: [
      suggestionPart.info,
      " ",
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_DocumentationLink.DocumentationLink, { suggestionPart })
    ] }),
    suggestionPart.synopsis && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_CodeSnippet.CodeSnippet,
      {
        language: "dql",
        showLineNumbers: false,
        showCopyAction: false,
        lineBreaks: true,
        className: (0, import_safe_sprinkles.globalSprinkles)({ marginTop: 4 }),
        children: suggestionPart.synopsis
      }
    )
  ] });
};
