// @dynatrace/strato-components/editors/dql-editor/dql/suggestion-autocompletions.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var suggestion_autocompletions_exports = {};
__export(suggestion_autocompletions_exports, {
  suggestionAutocompletion: () => suggestionAutocompletion
});
module.exports = __toCommonJS(suggestion_autocompletions_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_dql_linter = require("./dql-linter.js");
var import_fuzzy_match = require("./fuzzy-match.js");
var import_get_suggestions = require("./get-suggestions.js");
var import_SuggestionPart = require("./SuggestionPart.js");
var import_Flex = require("../../../layouts/flex/Flex.js");
var import_Text = require("../../../typography/text/Text.js");
function getMatchPattern(content) {
  const match = content.match(/\W?([\w.]*)$/);
  const capture = match?.[1];
  return capture ?? "";
}
function hasInfo(suggestions) {
  return suggestions.some(
    (suggestion) => suggestion.parts.some((part) => part.info !== void 0)
  );
}
function removeQuoteSuggestions(suggestions) {
  const excludedSuggestions = ['"', '""', '"""', " "];
  return suggestions.filter((elem) => !excludedSuggestions.includes(elem.suggestion)).map((elem) => {
    const trimmed = elem.suggestion.replace(/"{1,3}$/, "");
    return { ...elem, suggestion: trimmed };
  }).filter((elem) => elem.suggestion.length > 0);
}
function filterSuggestions(suggestions, pattern, requestCursorPosition, cursorPosition, disableFilter) {
  return suggestions.reduce((result, suggestion) => {
    const matchResult = (0, import_fuzzy_match.fuzzyMatch)(pattern, suggestion.suggestion);
    const match = suggestion.alreadyTypedCharacters === 0 ? true : matchResult.match;
    const score = suggestion.alreadyTypedCharacters === 0 ? 1 : matchResult.score;
    const startPos = suggestion.alreadyTypedCharacters === 0 ? cursorPosition : requestCursorPosition - suggestion.alreadyTypedCharacters;
    if (suggestion.suggestion.length > 0 && (disableFilter || match)) {
      result.push({ ...suggestion, score, startPos });
    }
    return result;
  }, []).sort((a, b) => b.score - a.score);
}
function getRequestString(autocompleteContent, patternLength, requestLength) {
  return autocompleteContent.slice(
    0,
    autocompleteContent.length - patternLength + requestLength
  );
}
const MAX_DATA_SUGGESTIONS = 1e4;
async function getFilteredSuggestions(autocompleteContent, patternLength, explicit) {
  const requestString = getRequestString(
    autocompleteContent,
    patternLength,
    explicit ? patternLength : 1
  );
  const suggestions = await (0, import_get_suggestions.getSuggestions)(
    requestString,
    requestString.length,
    MAX_DATA_SUGGESTIONS
  );
  const pattern = autocompleteContent.slice(
    autocompleteContent.length - patternLength
  );
  return filterSuggestions(
    suggestions,
    pattern,
    requestString.length,
    autocompleteContent.length,
    explicit
  );
}
const suggestionAutocompletion = async (context) => {
  if (context.matchBefore(/[^\w.]/) && !context.explicit) {
    return null;
  }
  const autocompleteContent = context.state.doc.slice(0, context.pos).toString();
  const pattern = getMatchPattern(autocompleteContent);
  const infoSuggestions = await (0, import_get_suggestions.getSuggestions)(
    getRequestString(autocompleteContent, pattern.length, 0)
  );
  let filteredSuggestions = await getFilteredSuggestions(
    autocompleteContent,
    pattern.length,
    pattern.length <= 1
  );
  if (filteredSuggestions.length === 0) {
    if (pattern.length > 1) {
      filteredSuggestions = await getFilteredSuggestions(
        autocompleteContent,
        pattern.length,
        true
      );
    }
  }
  if (context.tokenBefore(["string"]) !== null && autocompleteContent.length !== context.state.doc.length) {
    filteredSuggestions = removeQuoteSuggestions(filteredSuggestions);
  }
  if (filteredSuggestions.length === 0 && hasInfo(infoSuggestions)) {
    const emptySuggestion = infoSuggestions.find(
      (s) => s.suggestion.length === 0
    );
    if (emptySuggestion) {
      filteredSuggestions.push({
        ...emptySuggestion,
        score: -99,
        startPos: context.pos
      });
    }
  }
  if (filteredSuggestions.length === 0) {
    return null;
  }
  const autocompletePos = Math.min(
    ...filteredSuggestions.map((s) => s.startPos)
  );
  const section = infoSuggestions.filter((suggestion) => !suggestion.suggestion).flatMap((suggestion) => suggestion.parts.map((part) => part.info)).filter((text) => !!text).join(" ");
  return {
    suggestions: filteredSuggestions.map((suggestion) => {
      const hasInfo2 = suggestion.parts.some((part) => !!part.info);
      const shouldBeTruncated = (0, import_fuzzy_match.labelShouldBeTruncated)(suggestion.suggestion);
      const middleEllipsisText = shouldBeTruncated ? (0, import_fuzzy_match.convertToMiddleEllipsis)(suggestion.suggestion) : suggestion.suggestion;
      const infoPopup = hasInfo2 ? () => {
        return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Flex.Flex, { flexDirection: "column", gap: 16, children: suggestion.parts.filter((part) => !!part.info).map((part) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_SuggestionPart.SuggestionPart, { suggestionPart: part }, part.info)) });
      } : void 0;
      return {
        ...suggestion,
        label: middleEllipsisText,
        section,
        apply: (view) => {
          view.dispatch({
            changes: {
              from: suggestion.startPos,
              to: autocompleteContent.length,
              insert: suggestion.suggestion
            },
            selection: {
              anchor: suggestion.startPos + suggestion.suggestion.length
            },
            userEvent: "input.complete"
          });
        },
        popup: shouldBeTruncated ? () => /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { fontStyle: "code", as: "p", children: suggestion.suggestion }),
          infoPopup ?? null
        ] }) : infoPopup
      };
    }),
    from: autocompletePos,
    async reopen(completion) {
      const queryToEnd = context.state.doc.slice(context.pos).toString();
      const completedQuery = autocompleteContent.slice(0, completion.startPos) + completion.suggestion + queryToEnd;
      const { valid = true } = await (0, import_dql_linter.verifyQuery)(completedQuery) ?? {};
      return !valid;
    },
    filter: false,
    getMatch: (completion) => {
      return (0, import_fuzzy_match.getFuzzyMatchFullLabelPosition)(pattern, completion.suggestion) ?? (0, import_fuzzy_match.fuzzyMatch)(
        autocompleteContent.slice(autocompletePos),
        completion.suggestion ?? ""
      ).pos ?? [];
    }
  };
};
