// @dynatrace/strato-components/esm/charts/core/color-rule/strategies.js
import { get, isNil } from "lodash-es";
import { always0, getColorFromRule, getSubject, incBy1 } from "./utils.js";
import { getAccessor } from "../utils/getAccessor.js";
import { getValidColor } from "../utils/getValidColor.js";
import { RANGE_MATCHERS, VALUE_MATCHERS } from "../utils/validators.js";
const strategies = {
  histogram: () => {
    return {
      defaultValueAccessor: "name",
      forbiddenAccessors: /^name$/,
      getName: (series) => series.name,
      getSubject,
      matchers: VALUE_MATCHERS,
      colorPaletteIndex: incBy1,
      getColorFromRule
    };
  },
  timeseries: () => {
    return {
      defaultValueAccessor: "name",
      forbiddenAccessors: /^name$/,
      getName: (series) => Array.isArray(series.name) ? series.name[0] : series.name,
      getSubject,
      matchers: VALUE_MATCHERS,
      colorPaletteIndex: incBy1,
      getColorFromRule
    };
  },
  categoricalBar: () => {
    return {
      defaultValueAccessor: "category",
      forbiddenAccessors: /^category$/,
      getName: (series) => series.category,
      getSubject,
      matchers: VALUE_MATCHERS,
      colorPaletteIndex: always0,
      getColorFromRule: (rule) => {
        if (rule.colorPalette) {
          return rule.colorPalette;
        }
        return { Default: getValidColor(rule.color ?? "") };
      }
    };
  },
  meterBar: () => {
    const getValueWithAccessor = (currentSeries, valueAccessor) => {
      const accessor = getAccessor(valueAccessor);
      if (!accessor) {
        return null;
      }
      const value = get(currentSeries, accessor);
      return !isNil(value) ? value : null;
    };
    return {
      defaultValueAccessor: "value",
      forbiddenAccessors: /^value$/,
      getName: (series) => JSON.stringify(series.value),
      getSubject: function(rule, currentSeries) {
        const valueAccessor = rule.valueAccessor || currentSeries["_chartValueAccessor"];
        if (valueAccessor) {
          return getValueWithAccessor(currentSeries, valueAccessor);
        }
        return currentSeries.value;
      },
      matchers: VALUE_MATCHERS,
      colorPaletteIndex: incBy1,
      getColorFromRule
    };
  },
  pieOrDonut: () => {
    return {
      defaultValueAccessor: "value",
      forbiddenAccessors: /^(value|relativeValue)$/,
      getName: (series) => series.category,
      getSubject,
      matchers: VALUE_MATCHERS,
      colorPaletteIndex: incBy1,
      getColorFromRule
    };
  },
  honeycomb: () => {
    return {
      defaultValueAccessor: "value",
      forbiddenAccessors: /^value$/,
      getName: () => "honeycomb",
      getSubject,
      matchers: VALUE_MATCHERS,
      colorPaletteIndex: always0,
      getColorFromRule
    };
  },
  categoricalLegend: () => {
    return {
      defaultValueAccessor: "valueOf",
      forbiddenAccessors: /(?:)/,
      getName: () => "categoricalLegend",
      getSubject: (_rule, item) => item,
      matchers: VALUE_MATCHERS,
      colorPaletteIndex: always0,
      getColorFromRule
    };
  },
  numericalLegend: () => {
    return {
      defaultValueAccessor: "from",
      forbiddenAccessors: /(?:)/,
      getName: () => "numericalLegend",
      getSubject: (_, item) => item,
      matchers: RANGE_MATCHERS,
      colorPaletteIndex: always0,
      getColorFromRule
    };
  }
};
export {
  strategies
};
