// @dynatrace/strato-components/esm/charts/xy-chart/components/layout/XYChartGridLayout.js
import { jsx } from "react/jsx-runtime";
import { plotLayout } from "../../../core/components/chart-grid/ChartGrid.sty.js";
import { useGridPlotSize } from "../../../core/components/chart-grid/hooks/useGridPlotSize.js";
import { Grid } from "../../../core/components/grid/Grid.js";
import { isBandScale } from "../../../core/utils/detect-scale.js";
import { useAxesTicks } from "../../hooks/useAxesTicks.js";
import { useBarData } from "../../hooks/useBarData.js";
import { useGetVisibleAxesScales } from "../../hooks/useGetVisibleAxesScales.js";
import { useShouldShowGrid } from "../../hooks/useShouldShowGrid.js";
const XYChartGridLayout = () => {
  const { width, height } = useGridPlotSize();
  const { shouldShowGrid } = useShouldShowGrid();
  const visibleAxisScales = useGetVisibleAxesScales();
  const { topTicks, bottomTicks, leftTicks, rightTicks } = useAxesTicks();
  const { hasXGroups, hasYGroups } = useBarData();
  if (!shouldShowGrid()) {
    return null;
  }
  if (!visibleAxisScales.hasXAndYScale) {
    return null;
  }
  const xTicks = bottomTicks.numericTicks ?? topTicks.numericTicks;
  const yTicks = leftTicks.numericTicks ?? rightTicks.numericTicks;
  const xCategoryOffset = hasXGroups && isBandScale(visibleAxisScales.xScale) ? visibleAxisScales.xScale.step() / 2 : void 0;
  const yCategoryOffset = hasYGroups && isBandScale(visibleAxisScales.yScale) ? visibleAxisScales.yScale.step() / 2 : void 0;
  return /* @__PURE__ */ jsx("svg", { className: plotLayout, children: /* @__PURE__ */ jsx(
    Grid,
    {
      xScale: visibleAxisScales.xScale,
      yScale: visibleAxisScales.yScale,
      height,
      width,
      xTicks,
      yTicks,
      highlightZeroLine: {
        horizontal: false,
        vertical: false
      },
      xCategoryOffset,
      yCategoryOffset
    }
  ) });
};
XYChartGridLayout["displayName"] = "XYChartGridLayout";
export {
  XYChartGridLayout
};
