// @dynatrace/strato-components/esm/filters/filter-field/components/FilterStatement.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import { useMemo, Fragment, useRef, useEffect } from "react";
import { getOrCreateNamedHighlight } from "./CSSHighlight.js";
import * as filterfieldStyles from "../FilterField.sty.js";
import * as styles from "./FilterStatement.sty.js";
import { convertStringToFilterFieldTree } from "../utils/conversion/string-tree-conversion.js";
import {
  isFilterFieldBigIntNode,
  isFilterFieldBooleanNode,
  isFilterFieldComparisonOperatorNode,
  isFilterFieldDurationNode,
  isFilterFieldErrorNode,
  isFilterFieldExistsOperatorNode,
  isFilterFieldGroupNode,
  isFilterFieldInclusionOperatorNode,
  isFilterFieldJSONPathNode,
  isFilterFieldKeyNode,
  isFilterFieldLeafNode,
  isFilterFieldListNode,
  isFilterFieldNumberNode,
  isFilterFieldSearchOperatorNode,
  isFilterFieldStatementNode,
  isFilterFieldStringNode,
  isFilterFieldVariableNode
} from "../utils/tree-node-guards.js";
function WhitespaceRenderer(props) {
  const { start: startNode, end: endNode } = props;
  if (!startNode || !endNode) {
    return null;
  }
  const numberOfSpaces = Math.max(
    0,
    (endNode.range.from ?? -1) - (startNode.range.to ?? -1)
  );
  return Array(numberOfSpaces).fill(" ").map((item, index) => /* @__PURE__ */ jsx("span", { className: styles.whitespace, children: " " }, `${item}-${index}`));
}
function withNodeRenderer(className) {
  return function Component(props) {
    const { node } = props;
    const ref = useRef(null);
    const highlight = useMemo(
      () => getOrCreateNamedHighlight("filter-field-escape-characters"),
      []
    );
    useEffect(() => {
      const ranges = [];
      if (node?.textValue.includes("\\") && ref.current) {
        const textNode = ref.current.childNodes.item(0);
        if (textNode) {
          const textContent = textNode.textContent ?? "";
          let index = 0;
          while ((index = textContent.indexOf("\\", index)) !== -1) {
            if (index + 1 >= textContent.length) {
              break;
            }
            const range = new Range();
            range.setStart(textNode, index);
            range.setEnd(textNode, index + 2);
            ranges.push(range);
            index += 2;
            highlight.add(range);
          }
        }
      }
      return () => {
        for (const range of ranges) {
          highlight.delete(range);
        }
      };
    }, [highlight, node?.textValue]);
    return /* @__PURE__ */ jsx("span", { ref, className: clsx(styles.escapeCharacter, className), children: node?.textValue });
  };
}
const KeyRenderer = withNodeRenderer(
  filterfieldStyles.filterKey
);
const StringRenderer = withNodeRenderer(
  filterfieldStyles.filterValue
);
const ComparisonOperatorRenderer = withNodeRenderer(
  filterfieldStyles.operator
);
const ListRenderer = withNodeRenderer("");
const VariableRenderer = withNodeRenderer(
  filterfieldStyles.variable
);
const JSONPathRenderer = withNodeRenderer(
  filterfieldStyles.filterValue
);
const NumberRenderer = withNodeRenderer(
  filterfieldStyles.filterValue
);
const BigIntRenderer = withNodeRenderer(
  filterfieldStyles.filterValue
);
const BooleanRenderer = withNodeRenderer(
  filterfieldStyles.filterValue
);
const DurationRenderer = withNodeRenderer(
  filterfieldStyles.filterValue
);
const ExistsOperatorRenderer = withNodeRenderer(
  filterfieldStyles.operator
);
const SearchOperatorRenderer = withNodeRenderer(
  filterfieldStyles.operator
);
const InclusionOperatorRenderer = withNodeRenderer(
  filterfieldStyles.operator
);
const ErrorRenderer = withNodeRenderer(
  filterfieldStyles.error
);
function GroupRenderer(props) {
  const { node } = props;
  const lastChild = node.children[node.children.length - 1];
  return /* @__PURE__ */ jsxs("span", { children: [
    node.explicit ? "(" : "",
    node.children.map((childNode, index, allChildren) => {
      const isLast = index === allChildren.length - 1;
      const nextChild = node.children[index + 1];
      let joiner = null;
      if (node.explicitLogicalOperator && !isLast && nextChild?.type !== "Error" && (!("value" in nextChild) || nextChild.value !== "")) {
        joiner = /* @__PURE__ */ jsxs("span", { className: filterfieldStyles.operator, children: [
          /* @__PURE__ */ jsx("span", { className: styles.whitespace, children: " " }),
          node.logicalOperator,
          /* @__PURE__ */ jsx("span", { className: styles.whitespace, children: " " })
        ] });
      } else if (node.explicitLogicalOperator === false && !isLast) {
        joiner = /* @__PURE__ */ jsx(
          WhitespaceRenderer,
          {
            start: childNode,
            end: allChildren.at(index + 1)
          }
        );
      }
      return /* @__PURE__ */ jsxs(Fragment, { children: [
        /* @__PURE__ */ jsx(NodeRenderer, { node: childNode }),
        joiner
      ] }, `${childNode.range.from}-${childNode.range.to}`);
    }),
    node.explicit && lastChild.type !== "Error" && (!("value" in lastChild) || lastChild.value !== "") ? ")" : ""
  ] });
}
function ValueRenderer(props) {
  const { node } = props;
  if (isFilterFieldLeafNode(node)) {
    return /* @__PURE__ */ jsx(LeafNodeRenderer, { node });
  }
  if (isFilterFieldListNode(node)) {
    return /* @__PURE__ */ jsx(ListRenderer, { node });
  }
  return null;
}
function StatementRenderer(props) {
  const { node } = props;
  return /* @__PURE__ */ jsxs("span", { className: styles.statement, children: [
    /* @__PURE__ */ jsx(KeyRenderer, { node: node.key }),
    /* @__PURE__ */ jsx(WhitespaceRenderer, { start: node.key, end: node.comparisonOperator }),
    /* @__PURE__ */ jsx(LeafNodeRenderer, { node: node.comparisonOperator }),
    /* @__PURE__ */ jsx(WhitespaceRenderer, { start: node.comparisonOperator, end: node.value }),
    /* @__PURE__ */ jsx(ValueRenderer, { node: node.value })
  ] });
}
function LeafNodeRenderer(props) {
  const { node } = props;
  if (isFilterFieldKeyNode(node)) {
    return /* @__PURE__ */ jsx(KeyRenderer, { node });
  }
  if (isFilterFieldComparisonOperatorNode(node)) {
    return /* @__PURE__ */ jsx(ComparisonOperatorRenderer, { node });
  }
  if (isFilterFieldStringNode(node)) {
    return /* @__PURE__ */ jsx(StringRenderer, { node });
  }
  if (isFilterFieldVariableNode(node)) {
    return /* @__PURE__ */ jsx(VariableRenderer, { node });
  }
  if (isFilterFieldNumberNode(node)) {
    return /* @__PURE__ */ jsx(NumberRenderer, { node });
  }
  if (isFilterFieldJSONPathNode(node)) {
    return /* @__PURE__ */ jsx(JSONPathRenderer, { node });
  }
  if (isFilterFieldBigIntNode(node)) {
    return /* @__PURE__ */ jsx(BigIntRenderer, { node });
  }
  if (isFilterFieldBooleanNode(node)) {
    return /* @__PURE__ */ jsx(BooleanRenderer, { node });
  }
  if (isFilterFieldDurationNode(node)) {
    return /* @__PURE__ */ jsx(DurationRenderer, { node });
  }
  if (isFilterFieldExistsOperatorNode(node)) {
    return /* @__PURE__ */ jsx(ExistsOperatorRenderer, { node });
  }
  if (isFilterFieldSearchOperatorNode(node)) {
    return /* @__PURE__ */ jsx(SearchOperatorRenderer, { node });
  }
  if (isFilterFieldInclusionOperatorNode(node)) {
    return /* @__PURE__ */ jsx(InclusionOperatorRenderer, { node });
  }
  if (isFilterFieldErrorNode(node)) {
    return /* @__PURE__ */ jsx(ErrorRenderer, { node });
  }
  return null;
}
function NodeRenderer(props) {
  const { node } = props;
  if (isFilterFieldGroupNode(node)) {
    return /* @__PURE__ */ jsx(GroupRenderer, { node });
  }
  if (isFilterFieldStatementNode(node)) {
    return /* @__PURE__ */ jsx(StatementRenderer, { node });
  }
  if (isFilterFieldErrorNode(node)) {
    return /* @__PURE__ */ jsx(ErrorRenderer, { node });
  }
  return null;
}
function FilterStatement(props) {
  const {
    value: filterString,
    "data-testid": dataTestId,
    className: consumerClassName,
    style: consumerStyle,
    highlighting = "syntax"
  } = props;
  const tokenizedValue = useMemo(() => {
    return convertStringToFilterFieldTree(filterString ?? "", {
      parserConfig: {
        jsonPathConversion: true,
        searchConversion: true
      }
    });
  }, [filterString]);
  return /* @__PURE__ */ jsx(
    "div",
    {
      className: clsx(styles.root, consumerClassName, [
        highlighting === "syntax" && styles.highlighting.syntax,
        highlighting === "minimal" && styles.highlighting.minimal
      ]),
      style: consumerStyle,
      "data-testid": dataTestId,
      children: /* @__PURE__ */ jsx(NodeRenderer, { node: tokenizedValue.tree })
    }
  );
}
export {
  FilterStatement,
  NodeRenderer
};
