// @dynatrace/strato-components/esm/charts/pie/components/PieChartTooltip/PieChartTooltip.js
import { jsx } from "react/jsx-runtime";
import { OverlayTooltip } from "../../../core/components/overlay-tooltip/OverlayTooltip.js";
import { useSelector as useOverlayTooltipSelector } from "../../../core/components/overlay-tooltip/selectors.js";
import { isTooltipDataEmpty } from "../../../core/components/overlay-tooltip/utils.js";
import { useTooltipSeriesActions } from "../../../core/components/tooltip/shared/hooks/useTooltipSeriesActions.js";
import { useVisibleSlicesState } from "../../hooks/store/selectors.js";
import { usePieChartTooltip } from "../../hooks/usePieChartTooltip.js";
const PieChartTooltip = () => {
  const customTooltip = usePieChartTooltip();
  const overlayTooltipPinned = useOverlayTooltipSelector(
    (state) => state.pinned
  );
  const overlayData = useOverlayTooltipSelector((state) => state.data);
  const visibleSlices = useVisibleSlicesState();
  const seriesActions = useTooltipSeriesActions(
    visibleSlices,
    (slice) => slice.id
  );
  if (customTooltip?.hidden || isTooltipDataEmpty(overlayData)) {
    return null;
  }
  const metadata = overlayData.metadata;
  const currentSeries = overlayData.series[0];
  const fallbackSeriesId = typeof currentSeries.name === "string" ? currentSeries.name : "";
  if (customTooltip?.template) {
    const payload = {
      category: typeof currentSeries.name === "string" ? currentSeries.name : "",
      value: currentSeries.value,
      percentage: currentSeries.relativeValue,
      color: currentSeries.color
    };
    return /* @__PURE__ */ jsx(OverlayTooltip, { legacyTemplate: customTooltip.template(payload) });
  }
  const tooltipItem = currentSeries ? {
    id: metadata?.sliceId ?? fallbackSeriesId,
    name: currentSeries.name,
    color: { Default: currentSeries.color },
    values: [
      currentSeries.value?.toString() || "",
      currentSeries.relativeValue?.toString() || ""
    ]
  } : void 0;
  const actions = overlayTooltipPinned && tooltipItem ? () => seriesActions(tooltipItem) : void 0;
  return /* @__PURE__ */ jsx(OverlayTooltip, { children: /* @__PURE__ */ jsx(
    OverlayTooltip.SeriesItem,
    {
      name: currentSeries.name,
      value: currentSeries.value,
      color: currentSeries.color,
      relativeValue: currentSeries.relativeValue,
      relativeValueUnit: currentSeries.relativeValueUnit,
      actions
    }
  ) });
};
export {
  PieChartTooltip
};
