// @dynatrace/strato-components/esm/charts/core/components/tooltip/types/ChartTooltip.config.js
import { BubbleShape } from "../../shape/BubbleShape/BubbleShape.js";
import { EmojiShape } from "../../shape/EmojiShape/EmojiShape.js";
import { IconShape } from "../../shape/IconShape/IconShape.js";
import { MultiDataPointShape } from "../../shape/MultiDataPointShape.js";
import { SingleDataPointShape } from "../../shape/SingleDataPointShape.js";
import { ChipAtom } from "../../tooltip-atoms/components/ChipAtom.js";
import { DurationChipAtom } from "../../tooltip-atoms/components/DurationChipAtom/DurationChipAtom.js";
import { PrimaryAtom } from "../../tooltip-atoms/components/PrimaryAtom.js";
import { SecondaryAtom } from "../../tooltip-atoms/components/SecondaryAtom.js";
import { SymbolAtom } from "../../tooltip-atoms/components/SymbolAtom.js";
import { ActionItem } from "../ChartTooltip/components/ActionItem/ActionItem.js";
import { DurationChip } from "../ChartTooltip/components/DurationChip.js";
import { EventItem } from "../ChartTooltip/components/EventItem/EventItem.js";
import { GeoItem } from "../ChartTooltip/components/GeoItem.js";
import { KeyValue } from "../ChartTooltip/components/KeyValue.js";
import { RangeHeader } from "../ChartTooltip/components/RangeHeader/RangeHeader.js";
import { SeriesItem } from "../ChartTooltip/components/SeriesItem/SeriesItem.js";
import { TextHeader } from "../ChartTooltip/components/TextHeader.js";
import { TimeFrameHeader } from "../ChartTooltip/components/TimeFrameHeader/TimeFrameHeader.js";
import { TimestampHeader } from "../ChartTooltip/components/TimestampHeader/TimestampHeader.js";
import { TooltipText } from "../section-components/atoms/TooltipText.js";
import { TooltipBody } from "../section-components/TooltipBody.js";
import { TooltipFooter } from "../section-components/TooltipFooter/TooltipFooter.js";
import { TooltipHeader } from "../section-components/TooltipHeader/TooltipHeader.js";
import { TooltipInfoContent } from "../section-components/TooltipInfoContent/TooltipInfoContent.js";
import { TooltipItem } from "../section-components/TooltipItem/TooltipItem.js";
import { TooltipSubheader } from "../section-components/TooltipSubheader.js";
import { TooltipSymbolContent } from "../section-components/TooltipSymbolContent/TooltipSymbolContent.js";
import { TooltipValueContent } from "../section-components/TooltipValueContent/TooltipValueContent.js";
const ChartTooltipSlots = {
  // section components
  Header: TooltipHeader,
  Subheader: TooltipSubheader,
  Body: TooltipBody,
  Item: TooltipItem,
  Footer: TooltipFooter,
  // Item subsections
  Symbol: TooltipSymbolContent,
  Content: TooltipInfoContent,
  Value: TooltipValueContent
};
const ChartTooltipAtoms = {
  // atoms
  Chip: ChipAtom,
  DurationChipAtom,
  Primary: PrimaryAtom,
  Secondary: SecondaryAtom,
  Symbol: SymbolAtom,
  // shapes
  Bubble: BubbleShape,
  Emoji: EmojiShape,
  Icon: IconShape,
  MultiDataPoint: MultiDataPointShape,
  SingleDataPoint: SingleDataPointShape
};
const ChartTooltipTemplate = {
  GeoItem,
  Text: TooltipText,
  SeriesItem,
  ActionItem,
  TimeFrameHeader,
  TimestampHeader,
  RangeHeader,
  TextHeader,
  EventItem,
  KeyValue,
  DurationChip
};
export {
  ChartTooltipAtoms,
  ChartTooltipSlots,
  ChartTooltipTemplate
};
