// @dynatrace-sdk/user-preferences/testing/esm/index.js
/**
 * @license
 * Copyright 2023-2026 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// packages/platform/user-preferences/testing/test-utils.ts
var currentSetup;
var isInitialised = false;
var wrapper = {
  getLanguage: () => currentSetup.getLanguage(),
  getTheme: () => currentSetup.getTheme(),
  getRegionalFormat: () => currentSetup.getRegionalFormat(),
  getTimezone: () => currentSetup.getTimezone()
};
var setupTestingEnv = () => {
  if (typeof expect === "undefined") {
    isInitialised = true;
  }
  mockUserPreferences();
};
var mockUserPreferences = (setup = {}) => {
  if (!isInitialised) {
    throw new Error(`\`@dynatrace-sdk/user-preferences/testing\` has not been initialised!
Add in your \`jest.config.js\` file entry with \`@dynatrace-sdk/user-preferences/testing\` setup:

\`setupFiles: ['@dynatrace-sdk/user-preferences/testing']\``);
  }
  const globalWithDtRuntimeUserPreferences = globalThis;
  if (globalWithDtRuntimeUserPreferences.dtRuntime === void 0) {
    globalWithDtRuntimeUserPreferences.dtRuntime = {};
  }
  if (globalWithDtRuntimeUserPreferences.dtRuntime.userPreferences === void 0) {
    globalWithDtRuntimeUserPreferences.dtRuntime.userPreferences = wrapper;
  }
  const legacyDtSdk = globalThis;
  legacyDtSdk.dtSdk = legacyDtSdk.dtSdk || {};
  legacyDtSdk.dtSdk.getRegionalFormat = wrapper.getRegionalFormat;
  legacyDtSdk.dtSdk.getLanguage = wrapper.getLanguage;
  legacyDtSdk.dtSdk.getTheme = wrapper.getTheme;
  legacyDtSdk.dtSdk.getTimezone = wrapper.getTimezone;
  currentSetup = getSetupWithDefaults(setup);
};
var clearUserPreferencesMocks = () => {
  mockUserPreferences();
};
var getSetupWithDefaults = (setup) => {
  return {
    getTheme: setup.getTheme || (() => "auto"),
    getLanguage: setup.getLanguage || (() => "en"),
    getRegionalFormat: setup.getRegionalFormat || (() => "en"),
    getTimezone: setup.getTimezone || (() => "UTC")
  };
};

// packages/platform/user-preferences/testing/index.ts
setupTestingEnv();
export {
  clearUserPreferencesMocks,
  mockUserPreferences,
  setupTestingEnv
};
