// @dynatrace/strato-components/esm/charts/meter-bar/multi-meter-bar/components/MultiMeterBarChartSegments.js
import { Fragment, jsx } from "react/jsx-runtime";
import { MultiMeterBarChartSegment } from "./MultiMeterBarChartSegment.js";
import { useMeterBarLayoutSize } from "../../shared/context/MeterBarLayoutSizeContext.js";
import { getSegmentWidth } from "../../shared/utils/get-segment-width.js";
import { useMultiMeterBarVisibleSegmentsState } from "../hooks/store/selectors.js";
const MultiMeterBarChartSegments = (props) => {
  const { max, min } = props;
  const { segmentWidth: width } = useMeterBarLayoutSize();
  const segments = useMultiMeterBarVisibleSegmentsState();
  if (!segments || !segments.length) {
    return null;
  }
  const { segmentComponents } = segments.reduce(
    (acc, current, index) => {
      const segmentWidth = getSegmentWidth(current.value, min, max, width);
      const segmentComponent = /* @__PURE__ */ jsx(
        MultiMeterBarChartSegment,
        {
          previousSegmentsWidth: acc.usedWidth,
          segmentId: current.id,
          segmentIndex: index,
          segmentColor: current.color,
          segmentWidth
        },
        current.id
      );
      acc = {
        segmentComponents: [...acc.segmentComponents, segmentComponent],
        usedWidth: acc.usedWidth + segmentWidth
      };
      return acc;
    },
    {
      segmentComponents: [],
      usedWidth: 0
    }
  );
  return /* @__PURE__ */ jsx(Fragment, { children: segmentComponents });
};
MultiMeterBarChartSegments["displayName"] = "MultiMeterBarChartSegments";
export {
  MultiMeterBarChartSegments
};
