// @dynatrace/strato-components/esm/tables/DataTable/features/CellFormatter/CellFormatter.js
import { jsx } from "react/jsx-runtime";
import { memo } from "react";
import {
  adjustFractionDigits,
  format,
  formatCurrency,
  formatDate as formatDatetime,
  formatLong,
  units
} from "@dynatrace-sdk/units";
import {
  DATETIME_DEFAULT_FORMAT_OPTIONS,
  formatCurrencyDefault,
  formatDatetimeDefault
} from "../../../../core/utils/default-formatters.js";
import { devLog } from "../../../../core/utils/logging.js";
import { convertToNumberIfPureNumber } from "../../../../core/utils/string-utils.js";
const DATETIME_FORMAT_OPTIONS = {
  // e.g. "12/21/2024" for US English locale
  date: {
    year: "numeric",
    month: "numeric",
    day: "numeric"
  },
  // e.g. "5:45:30 PM" for US English locale
  time: {
    hour: "numeric",
    minute: "numeric",
    second: "numeric"
  },
  // e.g. "12/21/2024, 5:45:30 PM" for US English locale
  datetime: DATETIME_DEFAULT_FORMAT_OPTIONS
};
function isDatetimeShorthand(formatter) {
  return formatter === "date" || formatter === "time" || formatter === "datetime";
}
function resolveDatetimeFormatter(formatter) {
  if (!formatter) {
    return void 0;
  }
  if (isDatetimeShorthand(formatter)) {
    return DATETIME_FORMAT_OPTIONS[formatter];
  }
  return formatter;
}
const ColumnCellDatetimeFormatMemo = memo(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  (props) => {
    return props.value ? formatDatetime(Number(props.value), props.formatter) : "";
  }
);
const ColumnCellCurrencyFormatMemo = memo(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  (props) => {
    if (props.value === null || props.value === void 0 || isNaN(props.value)) {
      return "";
    }
    const { currency, ...options } = props.formatter ?? {};
    if (!currency) {
      return formatCurrencyDefault(Number(props.value), options);
    }
    return formatCurrency(Number(props.value), currency, options);
  }
);
const ColumnCellFormatMemo = memo(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  (props) => {
    return format(Number(props.value), props.formatter);
  }
);
const ColumnCellFormatLongMemo = memo(
  (props) => {
    return formatLong(props.value, props.formatter);
  }
);
function formatColumnCell(columnDef) {
  if (columnDef.columnType === "sparkline" || columnDef.columnType === "meterbar" || columnDef.columnType === "gantt" || columnDef.columnType === "log-content" || columnDef.columnType === "markdown") {
    return;
  }
  if (columnDef.columnType === "date" || columnDef.columnType === "datetime") {
    const resolvedFormatter = resolveDatetimeFormatter(columnDef.formatter);
    return ({ getValue }) => /* @__PURE__ */ jsx(
      ColumnCellDatetimeFormatMemo,
      {
        value: getValue(),
        formatter: resolvedFormatter
      }
    );
  }
  if (columnDef.columnType === "currency") {
    return ({ getValue }) => /* @__PURE__ */ jsx(
      ColumnCellCurrencyFormatMemo,
      {
        value: getValue(),
        formatter: columnDef.formatter
      }
    );
  }
  if (columnDef.columnType === "long") {
    return ({ getValue }) => {
      const value = getValue();
      if (typeof value === "string" || typeof value === "bigint" || typeof value === "number") {
        return /* @__PURE__ */ jsx(
          ColumnCellFormatLongMemo,
          {
            value,
            formatter: columnDef.formatter
          }
        );
      }
      return String(value);
    };
  }
  return ({ getValue }) => /* @__PURE__ */ jsx(ColumnCellFormatMemo, { value: getValue(), formatter: columnDef.formatter });
}
function applyColumnFormatters(columnDef) {
  if (columnDef.columnType === "sparkline" || columnDef.columnType === "meterbar" || columnDef.columnType === "gantt" || columnDef.columnType === "log-content" || columnDef.columnType === "markdown") {
    return {};
  }
  if (columnDef.formatter) {
    return {
      cell: formatColumnCell(columnDef)
    };
  }
  return {};
}
function formatCellValue(value, column) {
  if (column.columnType === "sparkline" || column.columnType === "meterbar" || column.columnType === "gantt" || column.columnType === "log-content" || column.columnType === "markdown") {
    return "";
  }
  if (value === null || value === void 0) {
    devLog(`The value '${value}' can not be formatted!`);
    return "";
  }
  const cleanedValue = typeof value === "number" || value instanceof Date ? value : Number.isNaN(+value) ? null : convertToNumberIfPureNumber(value);
  if (column.formatter && // DEPRECATION: [APPDEV-16044] BREAKING-CHANGE: Remove 'date' columnType.
  column.columnType !== "date" && column.columnType !== "datetime" && column.columnType !== "currency" && typeof cleanedValue === "number") {
    const formatOptions = column.formatter;
    return format(cleanedValue, formatOptions);
  }
  if (column.columnType === "bit" && typeof cleanedValue === "number") {
    const formatOptions = { input: units.data.bit };
    return format(cleanedValue, formatOptions);
  }
  if (column.columnType === "number" && typeof cleanedValue === "number") {
    const formatOptions = column.formatter;
    if (formatOptions) {
      return format(cleanedValue, formatOptions);
    }
    return adjustFractionDigits(cleanedValue);
  }
  if (column.columnType === "long" && (typeof value === "string" || typeof value === "bigint" || typeof value === "number")) {
    return formatLong(value, column.formatter);
  }
  if (
    // DEPRECATION: [APPDEV-16044] BREAKING-CHANGE: Remove 'date' columnType
    (column.columnType === "date" || column.columnType === "datetime") && (typeof cleanedValue === "number" || cleanedValue instanceof Date)
  ) {
    const resolvedFormatter = resolveDatetimeFormatter(column.formatter);
    if (resolvedFormatter) {
      return formatDatetime(cleanedValue, resolvedFormatter);
    } else {
      return formatDatetimeDefault(cleanedValue);
    }
  }
  if (column.columnType === "currency" && typeof cleanedValue === "number") {
    const formatOptions = column.formatter;
    const { currency, ...remainingOptions } = formatOptions ?? {};
    if (!currency) {
      return formatCurrencyDefault(cleanedValue, remainingOptions);
    }
    return formatCurrency(cleanedValue, currency, remainingOptions);
  }
  devLog(`The value '${value}' can not be formatted!`);
  return String(value);
}
export {
  applyColumnFormatters,
  formatCellValue
};
