// @dynatrace/strato-components/tables/DataTable/features/SubRows/SubRowCellContext.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SubRowCellContext_exports = {};
__export(SubRowCellContext_exports, {
  SubRowCellContextProvider: () => SubRowCellContextProvider,
  SubrowCellContext: () => SubrowCellContext
});
module.exports = __toCommonJS(SubRowCellContext_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
const SubrowCellContext = (0, import_react.createContext)({});
function SubRowCellContextProvider(props) {
  if (props.isSubRowIndicatorColumn) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      SubrowCellContext.Provider,
      {
        value: {
          isSubRowIndicatorColumn: props.isSubRowIndicatorColumn,
          isExpanded: props.cell.row.getIsExpanded(),
          canExpand: props.cell.row.getCanExpand(),
          canExpandDisabled: props.cell.row.getCanExpandDisabled()
        },
        children: props.children
      }
    );
  }
  return props.children;
}
