// @dynatrace/strato-components/tables/DataTable/features/SubRows/SubRowExpanderPath.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SubRowExpanderPath_exports = {};
__export(SubRowExpanderPath_exports, {
  SubRowExpanderPath: () => SubRowExpanderPath
});
module.exports = __toCommonJS(SubRowExpanderPath_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var styles = __toESM(require("./SubRows.sty.js"));
var import_constants = require("../../constants.js");
var import_DataTableContext = require("../../contexts/DataTableContext.js");
function generatePathFromConfig(config, height, verticalAlignment, estimatedRowHeight, rowDensity) {
  const spacer = import_constants.DATA_TABLE_SUB_ROW_SPACER;
  const rowDensityPadding = import_constants.DATA_TABLE_SUB_ROW_DENSITY_PADDING_MAP[rowDensity];
  let pathString = "";
  for (const [index, shape] of config.entries()) {
    const oneBaseIndex = index + 1;
    const x = oneBaseIndex * spacer - spacer / 2;
    let yLShape = height / 2;
    if (height > estimatedRowHeight) {
      const expanderSpace = rowDensityPadding + import_constants.DATA_TABLE_SUB_ROW_EXPANDER_ICON_SIZE / 2;
      if (verticalAlignment === "top") {
        yLShape = expanderSpace;
      } else if (verticalAlignment === "bottom") {
        yLShape = height - expanderSpace;
      }
    }
    const yTShape = -height + yLShape;
    if (shape === "Line") {
      pathString += `M ${x},0 V ${height} `;
    } else if (shape === "LShape") {
      pathString += `M ${x},0 V ${yLShape} h 200 `;
    } else if (shape === "TShape") {
      pathString += `M ${x},0 V ${height} m 0,${yTShape} h 25`;
    }
  }
  return pathString;
}
const InternalMemoizedSubRowExpanderPath = (props) => {
  const { rowId, width, rowLineConfig, rowCanExpand } = props;
  const table = (0, import_DataTableContext.useDataTable)();
  const estimatedRowHeight = table.estimateRowHeight({ index: 0 });
  const rowHeight = table.getRowHeight(rowId);
  const visualVariantConfig = table.getVisualVariantOptions();
  const rowDensity = visualVariantConfig.rowDensity;
  const svgWidth = rowCanExpand ? width : width + import_constants.DATA_TABLE_SUB_ROW_EXPANDER_ICON_SIZE;
  const verticalAlignment = table.getBodyVerticalAlignment();
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "svg",
    {
      className: (0, import_clsx.default)(
        styles.expandIndentSvg,
        styles.expandIndentSvgSize({ rowCanExpand })
      ),
      style: {
        width: svgWidth
      },
      preserveAspectRatio: "none",
      overflow: "visible",
      viewBox: `0 0 ${svgWidth} ${rowHeight}`,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        "path",
        {
          d: generatePathFromConfig(
            rowLineConfig,
            rowHeight,
            verticalAlignment,
            estimatedRowHeight,
            rowDensity
          ),
          fill: "transparent",
          vectorEffect: "non-scaling-stroke",
          strokeWidth: 1
        }
      )
    }
  );
};
const SubRowExpanderPath = (0, import_react.memo)(InternalMemoizedSubRowExpanderPath);
