// @dynatrace/strato-components/tables/DataTable/features/SubRows/SubRowExpanders.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SubRowExpanders_exports = {};
__export(SubRowExpanders_exports, {
  SubRowsCellExpander: () => SubRowsCellExpander,
  SubRowsHeaderExpander: () => SubRowsHeaderExpander
});
module.exports = __toCommonJS(SubRowExpanders_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
var import_strato_icons = require("@dynatrace/strato-icons");
var import_SubRowCellContext = require("./SubRowCellContext.js");
var import_SubRowExpanderPath = require("./SubRowExpanderPath.js");
var styles = __toESM(require("./SubRows.sty.js"));
var import_SubRows = require("./SubRows.js");
var import_constants = require("../../constants.js");
var import_DataTableContext = require("../../contexts/DataTableContext.js");
var import_messages = require("../../messages.js");
var import_column_types_types = require("../ColumnTypes/column-types-types.js");
function remeasureSubRowIndicatorColumn(table) {
  const subrowIndicatorColumnId = table.getSubRowIndicatorColumnId();
  if (subrowIndicatorColumnId) {
    table.remeasureColumn(subrowIndicatorColumnId);
  }
}
function SubRowsHeaderExpander(props) {
  const { isAllRowsExpanded } = props;
  const table = (0, import_DataTableContext.useDataTable)();
  const intl = (0, import_react_intl.useIntl)();
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "button",
    {
      type: "button",
      className: (0, import_clsx.default)(
        styles.expandButtonHeaderCellWrapper,
        styles.expanderButton
      ),
      onClick: () => {
        table.toggleAllRowsExpanded();
        remeasureSubRowIndicatorColumn(table);
      },
      "aria-label": intl.formatMessage(import_messages.dataTableMessages.subRowAllTrigger, {
        expandState: isAllRowsExpanded ? "other" : "expand"
      }),
      style: {
        color: import_colors.default.Text.Neutral.Subdued
      },
      "data-dt-name": "SubRowExpandCollapseAll",
      "data-dt-properties": `state:${isAllRowsExpanded ? "collapse" : "expand"}`,
      children: isAllRowsExpanded ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.MinusIcon, {}) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.PlusIcon, {})
    }
  );
}
function SubRowsCellExpander({
  columnType
}) {
  const intl = (0, import_react_intl.useIntl)();
  const table = (0, import_DataTableContext.useDataTable)();
  const row = (0, import_DataTableContext.useDataTableRow)();
  const subrowContext = (0, import_react.useContext)(import_SubRowCellContext.SubrowCellContext);
  const subRowExpandButtonWrapperRef = (0, import_react.useRef)(null);
  if (!row) {
    return null;
  }
  const {
    canExpand: rowCanExpand,
    canExpandDisabled: rowCanExpandDisabled,
    isExpanded
  } = subrowContext;
  const indentWidth = import_constants.DATA_TABLE_SUB_ROW_INDENT_GRID * row.depth;
  const shouldCorrectIndicatorWidth = (0, import_column_types_types.isChartColumnType)(columnType);
  const subRowDepth = table.getSubRowOpenDepth();
  const depthDifference = subRowDepth - row.depth;
  const widthCorrection = import_constants.DATA_TABLE_SUB_ROW_INDENT_GRID * depthDifference;
  const rowDensity = table.getVisualVariantOptions().rowDensity;
  let width = (0, import_SubRows.getSubRowCellWidth)(rowDensity, indentWidth);
  if (shouldCorrectIndicatorWidth) {
    width = `calc(${width} + ${widthCorrection}px)`;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    "div",
    {
      ref: subRowExpandButtonWrapperRef,
      className: (0, import_clsx.default)(styles.expandButtonCellWrapper),
      style: {
        width
      },
      children: [
        row.depth > 0 ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_SubRowExpanderPath.SubRowExpanderPath,
          {
            rowId: row.id,
            width: indentWidth,
            rowLineConfig: table.getRowLineConfig(row.id),
            rowCanExpand
          }
        ) : null,
        rowCanExpand ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          "button",
          {
            disabled: rowCanExpandDisabled,
            type: "button",
            "aria-label": intl.formatMessage(import_messages.dataTableMessages.subRowTrigger, {
              expandState: isExpanded ? "other" : "expand",
              id: row.id
            }),
            "aria-expanded": isExpanded,
            className: styles.expanderButton,
            onClick: () => {
              row.toggleExpanded(!isExpanded);
              remeasureSubRowIndicatorColumn(table);
            },
            "data-dt-name": "SubRowExpandCollapse",
            "data-dt-properties": `state:${isExpanded ? "collapse" : "expand"}`,
            children: isExpanded ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.MinusIcon, { size: import_constants.DATA_TABLE_SUB_ROW_EXPANDER_ICON_SIZE }) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.PlusIcon, { size: import_constants.DATA_TABLE_SUB_ROW_EXPANDER_ICON_SIZE })
          }
        ) : null,
        shouldCorrectIndicatorWidth ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { style: { width: widthCorrection } }) : null
      ]
    }
  );
}
