// @dynatrace/strato-components/forms/spin-buttons/text-spin-button/TextSpinButton.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TextSpinButton_exports = {};
__export(TextSpinButton_exports, {
  TextSpinButton: () => TextSpinButton,
  isTextSpinButton: () => isTextSpinButton
});
module.exports = __toCommonJS(TextSpinButton_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_is = require("react-is");
var import_useValueNavigation = require("./hooks/useValueNavigation.js");
var import_useAriaLabelingProps = require("../../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../../core/hooks/useBehavioralTrackingProps.js");
var import_useSelection = require("../hooks/useSelection.js");
var import_useSpinButtonInteraction = require("../hooks/useSpinButtonInteraction.js");
var import_SpinButton_css = require("../SpinButton.sty.js");
function isTextSpinButton(node) {
  return node !== void 0 && node !== null && (0, import_react_is.isElement)(node) && node.type === TextSpinButton;
}
const TextSpinButton = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    placeholder,
    disabled = false,
    readOnly = false,
    options,
    required = false,
    hasError,
    className: consumerClassName,
    style: consumerStyle,
    onKeyDownHandler: propOnKeyDownHandler,
    "data-testid": dataTestId,
    onChange,
    "data-pattern": dataPattern,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const ariaProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(props);
  const rootRef = (0, import_react.useRef)(null);
  const { getSelection, setSelection, selectAll, deselectAll } = (0, import_useSelection.useSelection)(rootRef);
  const [error, setError] = (0, import_react.useState)(false);
  const { value, onKeyDownHandler, searchValue, clearValue } = (0, import_useValueNavigation.useValueNavigation)(props);
  (0, import_react.useImperativeHandle)(
    forwardedRef,
    () => {
      return {
        type: "text",
        element: rootRef,
        setSelection,
        getSelection,
        selectAll,
        deselectAll,
        searchValue,
        clearValue,
        onChange
      };
    },
    [
      clearValue,
      deselectAll,
      getSelection,
      onChange,
      searchValue,
      selectAll,
      setSelection
    ]
  );
  (0, import_react.useEffect)(() => {
    if (!rootRef.current) {
      return;
    }
    rootRef.current.textContent = value ?? "";
    document.activeElement === rootRef.current && selectAll();
  }, [value]);
  (0, import_react.useEffect)(() => {
    if (required && value === null) {
      setError(true);
    } else {
      setError(value !== null && !options.includes(value));
    }
  }, [required, value, options]);
  const { onClick, onFocus, onBlur } = (0, import_useSpinButtonInteraction.useSpinButtonInteraction)({
    disabled,
    selectAll,
    deselectAll,
    rootElement: rootRef.current
  });
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "span",
    {
      "data-testid": dataTestId,
      contentEditable: (
        // contentEditable when focused and non-editable when blurred, allowing text selection.
        !disabled && !readOnly ? true : void 0
      ),
      ref: rootRef,
      tabIndex: disabled ? void 0 : 0,
      role: "spinbutton",
      spellCheck: false,
      className: (0, import_clsx.default)(
        consumerClassName,
        (0, import_SpinButton_css.spinButtonCSS)({ disabled, error: hasError ?? error })
      ),
      style: {
        "--data-placeholder": `"${placeholder}"`,
        ...consumerStyle
      },
      "aria-disabled": disabled,
      ...ariaProps,
      "aria-label": ariaProps["aria-label"] ?? placeholder,
      onKeyDown: propOnKeyDownHandler ?? onKeyDownHandler,
      onClick,
      onFocus,
      onBlur,
      onBeforeInput: (e) => e.preventDefault(),
      onPaste: (e) => e.preventDefault(),
      onDrop: (e) => e.preventDefault(),
      onCut: (e) => e.preventDefault(),
      "data-pattern": dataPattern,
      ...behavioralTrackingProps
    }
  );
});
TextSpinButton.displayName = "_TextSpinButton";
