// @dynatrace/strato-components/esm/layouts/title-bar/TitleBar.js
import { jsx, jsxs } from "react/jsx-runtime";
import { forwardRef } from "react";
import Spacings from "@dynatrace/strato-design-tokens/spacings";
import { Action } from "./Action.js";
import { Navigation } from "./Navigation.js";
import { Prefix } from "./Prefix.js";
import { Subtitle } from "./Subtitle.js";
import { Suffix } from "./Suffix.js";
import { Title } from "./Title.js";
import { titleSubtitleWrapperCSS } from "./TitleBar.sty.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useSlots } from "../../core/hooks/useSlots.js";
import { canReceiveDefaultTextFormat as _canReceiveDefaultTextFormat } from "../../core/utils/can-receive-default-text-format.js";
import { Divider } from "../divider/Divider.js";
import { Flex } from "../flex/Flex.js";
const COMPONENT_NAME = "TitleBar";
const TitleBarRoot = /* @__PURE__ */ forwardRef(
  (props, forwardedRef) => {
    const {
      children,
      showDivider = true,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...remainingProps
    } = props;
    const [slots] = useSlots(children, {
      title: Title,
      subtitle: Subtitle,
      navigation: Navigation,
      suffix: Suffix,
      prefix: Prefix,
      action: Action
    });
    const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
    const titleIsText = !!slots.title && _canReceiveDefaultTextFormat(slots.title.props.children);
    const subtitleIsText = !!slots.subtitle && _canReceiveDefaultTextFormat(slots.subtitle?.props.children);
    const titleAndSubtitleIsText = titleIsText && !slots.subtitle || titleIsText && subtitleIsText;
    return /* @__PURE__ */ jsxs(
      Flex,
      {
        ref: forwardedRef,
        ...restProps,
        ...behavioralTrackingProps,
        width: "100%",
        flexDirection: "column",
        gap: 0,
        className: consumerClassName,
        style: consumerStyle,
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        "data-dt-component": COMPONENT_NAME,
        children: [
          /* @__PURE__ */ jsxs(Flex, { flexDirection: "column", gap: 8, children: [
            slots.navigation && /* @__PURE__ */ jsxs(Flex, { width: "100%", justifyContent: "space-between", gap: 8, children: [
              slots.navigation,
              slots.action
            ] }),
            /* @__PURE__ */ jsxs(Flex, { width: "100%", justifyContent: "space-between", gap: 8, children: [
              /* @__PURE__ */ jsxs(Flex, { flexWrap: "wrap", gap: 8, width: "100%", minWidth: 0, children: [
                slots.prefix,
                (slots.title || slots.subtitle) && /* @__PURE__ */ jsxs(
                  "div",
                  {
                    className: titleSubtitleWrapperCSS({
                      childrenAreText: titleAndSubtitleIsText
                    }),
                    children: [
                      slots.title,
                      slots.subtitle
                    ]
                  }
                ),
                slots.suffix
              ] }),
              !slots.navigation && slots.action
            ] })
          ] }),
          showDivider && /* @__PURE__ */ jsx(Divider, { flexItem: true, style: { marginTop: Spacings.Size8 } })
        ]
      }
    );
  }
);
TitleBarRoot.displayName = "TitleBar";
const TitleBar = Object.assign(TitleBarRoot, {
  Title,
  Subtitle,
  Suffix,
  Navigation,
  Prefix,
  Action
});
export {
  TitleBar
};
