// @dynatrace/strato-components/charts/xy-chart/components/overlay/explore/XYChartRectanglePoint.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var XYChartRectanglePoint_exports = {};
__export(XYChartRectanglePoint_exports, {
  XYChartRectanglePoint: () => XYChartRectanglePoint
});
module.exports = __toCommonJS(XYChartRectanglePoint_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_HighlightedRectangleCanvas = require("./HighlightedRectangleCanvas.js");
var import_XYChartExplore_css = require("./XYChartExplore.sty.js");
var import_Canvas = require("../../../../core/components/canvas/Canvas.js");
var import_CanvasContext = require("../../../../core/components/canvas/CanvasContext.js");
var import_useGridPlotSize = require("../../../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_defaults = require("../../../constants/defaults.js");
var import_useBarData = require("../../../hooks/useBarData.js");
var import_useRectData = require("../../../hooks/useRectData.js");
var import_useXYChartData = require("../../../hooks/useXYChartData.js");
var import_useXYChartRectSelection = require("../../../hooks/useXYChartRectSelection.js");
var import_useXYChartScales = require("../../../hooks/useXYChartScales.js");
var import_selectors = require("../../../store/selectors.js");
var import_useTooltipMetadata = require("../hooks/useTooltipMetadata.js");
var import_get_selected_datapoints = require("../utils/get-selected-datapoints.js");
const XYChartRectanglePoint = (0, import_react.memo)(() => {
  const contentState = (0, import_useTooltipMetadata.useTooltipMetadata)("rectangle");
  const selectState = (0, import_selectors.useSelectState)();
  const { width, height } = (0, import_useGridPlotSize.useGridPlotSize)();
  const axisIdToScales = (0, import_useXYChartScales.useXYChartScales)();
  const rectSelection = (0, import_useXYChartRectSelection.useXYChartRectSelection)();
  const { datapointMetadataMap } = (0, import_useXYChartData.useXYChartData)();
  const { axisIdToScaleDomain, axes } = (0, import_useXYChartData.useXYChartData)();
  const {
    seriesToMetadata: barSeriesMeta,
    filteredSeriesWithDatapoints: barSeriesWithDatapoints
  } = (0, import_useBarData.useBarData)();
  const {
    seriesToMetadata: rectSeriesMeta,
    filteredSeriesWithDatapoints: rectSeriesWithDatapoints
  } = (0, import_useRectData.useRectData)();
  const highlightedRects = (0, import_get_selected_datapoints.buildRectangleHighlightedPoints)({
    contentState,
    selectState,
    rectSelection,
    width,
    rectSeriesWithDatapoints,
    barSeriesWithDatapoints,
    axisIdToScales,
    barSeriesMeta,
    rectSeriesMeta,
    datapointMetadataMap,
    axisIdToScaleDomain,
    axes
  });
  if (!highlightedRects.length) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: import_XYChartExplore_css.datapointIndicatorLayout, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_CanvasContext.CanvasContext.Provider,
    {
      value: {
        shapeRefs: /* @__PURE__ */ new Map()
      },
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_Canvas.Canvas,
        {
          width: width + import_defaults.RECT_STROKE_WIDTH,
          height: height + import_defaults.RECT_STROKE_WIDTH,
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_HighlightedRectangleCanvas.HighlightedRectangleCanvas, { points: highlightedRects })
        }
      )
    }
  ) });
});
XYChartRectanglePoint.displayName = "XYChartRectanglePoint";
