// @dynatrace/strato-components/charts/xy-chart/components/overlay/explore/XYChartCrosshair.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var XYChartCrosshair_exports = {};
__export(XYChartCrosshair_exports, {
  XYChartCrosshair: () => XYChartCrosshair
});
module.exports = __toCommonJS(XYChartCrosshair_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_XYChartCrosshair_css = require("./XYChartCrosshair.sty.js");
var import_ChartGrid_css = require("../../../../core/components/chart-grid/ChartGrid.sty.js");
var import_useGridPlotSize = require("../../../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_HorizontalCrosshair = require("../../../../core/components/crosshair/HorizontalCrosshair.js");
var import_VerticalCrosshair = require("../../../../core/components/crosshair/VerticalCrosshair.js");
var import_selectors = require("../../../../core/components/overlay-tooltip/selectors.js");
var import_useTooltipReducer = require("../../../../core/components/overlay-tooltip/useTooltipReducer.js");
var import_useApplyCallbackOnScroll = require("../../../../core/hooks/useApplyCallbackOnScroll.js");
var import_useAxesFormatter = require("../../../hooks/useAxesFormatter.js");
var import_useFormatCategory = require("../../../hooks/useFormatCategory.js");
var import_useGetThresholds = require("../../../hooks/useGetThresholds.js");
var import_useVisibleAxesConfig = require("../../../hooks/useVisibleAxesConfig.js");
var import_useVisibleAxesScales = require("../../../hooks/useVisibleAxesScales.js");
var import_useXYChartData = require("../../../hooks/useXYChartData.js");
var import_useXYChartScales = require("../../../hooks/useXYChartScales.js");
var import_selectors2 = require("../../../store/selectors.js");
var import_state = require("../../../types/state.js");
var import_type_guards = require("../../../utils/type-guards.js");
var import_value_representation = require("../../../utils/value-representation.js");
var import_calculate_crosshair_coordinates = require("../utils/calculate-crosshair-coordinates.js");
var import_chartCrosshairDataTransformer = require("../utils/chartCrosshairDataTransformer.js");
function XYChartCrosshairLines({
  getLabels,
  height,
  thresholdOffset,
  thresholdYAxes,
  axisIdToScaleDomain,
  seriesToMetadata,
  datapointMetadataMap,
  axisIdToScales,
  hasVisibleSeriesOnThresholdAxis,
  chartBounds,
  isBottomLabelRotated,
  isTopLabelRotated,
  widthWithBaseOffset
}) {
  (0, import_selectors.useSelector)(({ relativeX: relativeX2, relativeY: relativeY2 }) => `${relativeX2}-${relativeY2}`);
  const { relativeX, relativeY, pinned, data } = (0, import_selectors.useTooltipSnapshot)();
  const position = relativeX !== null && relativeY !== null ? { x: relativeX, y: relativeY } : void 0;
  const tooltipContent = data.metadata || import_state.UndefinedContent;
  const { xAxis, yAxis, variant, orientation, isStacked } = seriesToMetadata.get(tooltipContent?.series) ?? {};
  const isArea = (0, import_type_guards.isAreaTooltipContent)(tooltipContent);
  const isHorizontal = orientation === "horizontal";
  const calculatedCoordinates = (0, import_react.useMemo)(() => {
    const tooltipContentHasCenteredCrosshair = (0, import_type_guards.isDotTooltipContent)(tooltipContent) || (0, import_type_guards.isLineTooltipContent)(tooltipContent) || (0, import_type_guards.isAreaTooltipContent)(tooltipContent);
    if (!pinned || !tooltipContentHasCenteredCrosshair || !tooltipContent.series || !tooltipContent.datapoint) {
      return null;
    }
    const xRelAccessor = xAxis ? (0, import_value_representation.relAccessor)(xAxis, axisIdToScaleDomain) : void 0;
    const yRelAccessor = yAxis ? (0, import_value_representation.relAccessor)(yAxis, axisIdToScaleDomain) : void 0;
    return (0, import_calculate_crosshair_coordinates.calculateCrosshairCoordinates)({
      pinned,
      tooltipContentHasCenteredCrosshair,
      tooltipContentSeries: tooltipContent.series,
      datapoint: tooltipContent.datapoint,
      axisIdToScales,
      datapointMetadataMap,
      xRelAccessor,
      yRelAccessor,
      isArea,
      isHorizontal,
      isStacked
    });
  }, [
    tooltipContent,
    xAxis,
    yAxis,
    axisIdToScaleDomain,
    pinned,
    axisIdToScales,
    datapointMetadataMap,
    isArea,
    isHorizontal,
    isStacked
  ]);
  const coordinates = calculatedCoordinates ?? position ?? { x: 0, y: 0 };
  if (!position) {
    return null;
  }
  const labels = getLabels({ position, pinned });
  const leftOffset = labels.left && hasVisibleSeriesOnThresholdAxis(thresholdYAxes.left) ? thresholdOffset.left : 0;
  const rightOffset = labels.right && hasVisibleSeriesOnThresholdAxis(thresholdYAxes.right) ? thresholdOffset.right : 0;
  const widthWithOffset = widthWithBaseOffset + leftOffset + rightOffset;
  const isBand = variant === "band";
  const showHorizontalCrosshair = !isBand || isHorizontal || !pinned;
  const showVerticalCrosshair = !isBand || !isHorizontal || !pinned;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("svg", { className: (0, import_clsx.default)(import_ChartGrid_css.plotLayout, import_XYChartCrosshair_css.crosshair), children: [
    showHorizontalCrosshair && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_HorizontalCrosshair.HorizontalCrosshair,
      {
        dashed: !pinned,
        leftLabel: labels.left,
        rightLabel: labels.right,
        y: coordinates.y,
        width: widthWithOffset,
        height,
        chartBounds,
        leftOffset,
        rightOffset
      }
    ),
    showVerticalCrosshair && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_VerticalCrosshair.VerticalCrosshair,
      {
        dashed: !pinned,
        bottomLabel: labels.bottom,
        topLabel: labels.top,
        isBottomLabelRotated,
        isTopLabelRotated,
        x: coordinates.x,
        height,
        width: widthWithOffset,
        chartBounds
      }
    )
  ] });
}
function XYChartCrosshair({
  chartRef
}) {
  const filteredSeriesIds = (0, import_selectors2.useFilteredSeriesIdsState)();
  const axesScale = (0, import_useVisibleAxesScales.useVisibleAxesScales)();
  const axesConfig = (0, import_useVisibleAxesConfig.useVisibleAxesConfig)();
  const axesIdToFormatter = (0, import_useAxesFormatter.useAxesFormatter)();
  const formatCategory = (0, import_useFormatCategory.useFormatCategory)();
  const { width, height } = (0, import_useGridPlotSize.useGridPlotSize)();
  const { offset: thresholdOffset, yAxes: thresholdYAxes } = (0, import_useGetThresholds.useGetThresholds)();
  const { axisIdToScaleDomain, seriesToMetadata, datapointMetadataMap } = (0, import_useXYChartData.useXYChartData)();
  const axisIdToScales = (0, import_useXYChartScales.useXYChartScales)();
  const chartBounds = chartRef.current?.getBoundingClientRect();
  const dispatch = (0, import_useTooltipReducer.useTooltipReducer)();
  const pinned = (0, import_selectors.useSelector)(({ pinned: pinned2 }) => pinned2);
  (0, import_useApplyCallbackOnScroll.useApplyCallbackOnScroll)(pinned, dispatch({ type: "RESET_TOOLTIP" }));
  const isBottomLabelRotated = axesConfig["bottom"]?.tickLabelLayout === "vertical";
  const isTopLabelRotated = axesConfig["top"]?.tickLabelLayout === "vertical";
  const hasVisibleSeriesOnThresholdAxis = (() => {
    if (!filteredSeriesIds) {
      return (_axisIds) => true;
    }
    const visibleSeriesEntries = [...seriesToMetadata.entries()].filter(
      ([series]) => filteredSeriesIds.has(series.id)
    );
    return (axisIds) => {
      return visibleSeriesEntries.some(
        ([, series]) => axisIds.has(series.yAxis?.id ?? "")
      );
    };
  })();
  const getLabels = (0, import_chartCrosshairDataTransformer.chartCrosshairDataTransformer)({
    axesScale,
    axesConfig,
    axesIdToFormatter,
    width,
    formatCategory
  });
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    XYChartCrosshairLines,
    {
      getLabels,
      height,
      thresholdOffset,
      thresholdYAxes,
      axisIdToScaleDomain,
      seriesToMetadata,
      datapointMetadataMap,
      axisIdToScales,
      hasVisibleSeriesOnThresholdAxis,
      chartBounds,
      isBottomLabelRotated,
      isTopLabelRotated,
      widthWithBaseOffset: width
    }
  );
}
