// @dynatrace/strato-components/charts/xy-chart/components/overlay/explore/XYChartLinearPoint.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var XYChartLinearPoint_exports = {};
__export(XYChartLinearPoint_exports, {
  XYChartLinearPoint: () => XYChartLinearPoint
});
module.exports = __toCommonJS(XYChartLinearPoint_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_HighlightedPointCanvas = require("./HighlightedPointCanvas.js");
var import_XYChartExplore_css = require("./XYChartExplore.sty.js");
var import_Canvas = require("../../../../core/components/canvas/Canvas.js");
var import_CanvasContext = require("../../../../core/components/canvas/CanvasContext.js");
var import_useGridPlotSize = require("../../../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_constants = require("../../../../core/components/data-point/constants.js");
var import_useAreaData = require("../../../hooks/useAreaData.js");
var import_useDotData = require("../../../hooks/useDotData.js");
var import_useLineData = require("../../../hooks/useLineData.js");
var import_useXYChartData = require("../../../hooks/useXYChartData.js");
var import_useXYChartLinearSelection = require("../../../hooks/useXYChartLinearSelection.js");
var import_useXYChartScales = require("../../../hooks/useXYChartScales.js");
var import_selectors = require("../../../store/selectors.js");
var import_useTooltipMetadata = require("../hooks/useTooltipMetadata.js");
var import_get_selected_datapoints = require("../utils/get-selected-datapoints.js");
const XYChartLinearPoint = (0, import_react.memo)(() => {
  const contentState = (0, import_useTooltipMetadata.useTooltipMetadata)("linear");
  const selectState = (0, import_selectors.useSelectState)();
  const { width, height } = (0, import_useGridPlotSize.useGridPlotSize)();
  const axisIdToScales = (0, import_useXYChartScales.useXYChartScales)();
  const linearSelection = (0, import_useXYChartLinearSelection.useXYChartLinearSelection)();
  const { axisIdToScaleDomain } = (0, import_useXYChartData.useXYChartData)();
  const { axes, datapointMetadataMap } = (0, import_useXYChartData.useXYChartData)();
  const {
    seriesToMetadata: dotSeriesMeta,
    filteredSeriesWithDatapoints: dotSeriesWithDatapoints
  } = (0, import_useDotData.useDotData)();
  const {
    seriesToMetadata: lineSeriesMeta,
    filteredSeriesWithDatapoints: lineSeriesWithDatapoints
  } = (0, import_useLineData.useLineData)();
  const {
    seriesToMetadata: areaSeriesMeta,
    filteredSeriesWithDatapoints: areaSeriesWithDatapoints
  } = (0, import_useAreaData.useAreaData)();
  const showContentPoint = (0, import_get_selected_datapoints.hasLinearContentPoint)(contentState);
  const showPoint = !showContentPoint && (!selectState || linearSelection.selectionConfigByAxisId.size === 0);
  if (showPoint) {
    return null;
  }
  const selectionPoints = (0, import_get_selected_datapoints.buildSelectionHighlightedPoints)({
    selectState,
    linearSelection,
    width,
    axisIdToScales,
    dotSeriesWithDatapoints,
    lineSeriesWithDatapoints,
    areaSeriesWithDatapoints,
    dotSeriesMeta,
    lineSeriesMeta,
    areaSeriesMeta,
    axisIdToScaleDomain,
    axes,
    datapointMetadataMap
  });
  const contentPoints = showContentPoint ? (0, import_get_selected_datapoints.buildContentHighlightedPoint)({
    contentState,
    axisIdToScales,
    dotSeriesMeta,
    lineSeriesMeta,
    areaSeriesMeta,
    axisIdToScaleDomain,
    datapointMetadataMap
  }) : [];
  const highlightedPoints = [
    ...selectionPoints,
    ...contentPoints
  ];
  if (!highlightedPoints.length) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: import_XYChartExplore_css.datapointIndicatorShiftedLayout, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_CanvasContext.CanvasContext.Provider,
    {
      value: {
        shapeRefs: /* @__PURE__ */ new Map()
      },
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_Canvas.Canvas,
        {
          width: width + 2 * import_constants.POINT_OVERLAY_PADDING,
          height: height + 2 * import_constants.POINT_OVERLAY_PADDING,
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_HighlightedPointCanvas.HighlightedPointCanvas, { points: highlightedPoints })
        }
      )
    }
  ) });
});
XYChartLinearPoint["displayName"] = "XYChartLinearPoint";
