// @dynatrace/strato-components/charts/categorical-bar/components/axes/CategoricalBarChartValueAxis.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CategoricalBarChartValueAxis_exports = {};
__export(CategoricalBarChartValueAxis_exports, {
  CategoricalBarChartValueAxis: () => CategoricalBarChartValueAxis
});
module.exports = __toCommonJS(CategoricalBarChartValueAxis_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_ValueBottomAxis = require("../../../core/components/axes/components/ValueBottomAxis.js");
var import_ValueLeftAxis = require("../../../core/components/axes/components/ValueLeftAxis.js");
var import_ChartGridRoot = require("../../../core/components/chart-grid/ChartGridRoot.js");
var import_useCategoricalBarChartBaseConfig = require("../../hooks/useCategoricalBarChartBaseConfig.js");
var import_useCategoricalBarChartScales = require("../../hooks/useCategoricalBarChartScales.js");
var import_useCategoricalBarChartTicks = require("../../hooks/useCategoricalBarChartTicks.js");
var import_useCategoricalBarChartValueAxisConfig = require("../../hooks/useCategoricalBarChartValueAxisConfig.js");
var import_categorical_bar_area_rules = require("../../types/categorical-bar-area-rules.js");
var import_layout = require("../../utils/layout.js");
const CategoricalBarChartValueAxis = () => {
  const { valueScale } = (0, import_useCategoricalBarChartScales.useCategoricalBarChartScales)();
  const { label: valueAxisLabel } = (0, import_useCategoricalBarChartValueAxisConfig.useCategoricalBarChartValueAxisConfig)();
  const { layout } = (0, import_useCategoricalBarChartBaseConfig.useCategoricalBarChartBaseConfig)();
  const { valueAxis } = (0, import_useCategoricalBarChartTicks.useCategoricalBarChartTicks)();
  return (0, import_layout.isVertical)(layout) ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ChartGridRoot.ChartGrid.GridArea, { gridArea: import_categorical_bar_area_rules.AREAS.leftAxis, children: ({ availableWidth, availableHeight }) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ValueLeftAxis.ValueLeftAxis,
    {
      formattedTicks: valueAxis.formattedTicks,
      ticks: valueAxis.ticks,
      containerHeight: availableHeight,
      containerWidth: availableWidth,
      scale: valueScale,
      label: valueAxisLabel
    }
  ) }) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ChartGridRoot.ChartGrid.GridArea, { gridArea: import_categorical_bar_area_rules.AREAS.bottomAxis, children: ({ availableWidth, availableHeight }) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ValueBottomAxis.ValueBottomAxis,
    {
      formattedTicks: valueAxis.formattedTicks,
      ticks: valueAxis.ticks,
      containerHeight: availableHeight,
      containerWidth: availableWidth,
      scale: valueScale,
      label: valueAxisLabel
    }
  ) });
};
CategoricalBarChartValueAxis["displayName"] = "CategoricalBarChartValueAxis";
