// @dynatrace/strato-components/charts/categorical-bar/components/axes/CategoricalBarChartCategoryAxis.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CategoricalBarChartCategoryAxis_exports = {};
__export(CategoricalBarChartCategoryAxis_exports, {
  CategoricalBarChartCategoryAxis: () => CategoricalBarChartCategoryAxis
});
module.exports = __toCommonJS(CategoricalBarChartCategoryAxis_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_categorical_category_axis = require("./utils/categorical-category-axis.js");
var import_CategoryBottomAxis = require("../../../core/components/axes/components/CategoryBottomAxis.js");
var import_CategoryLeftAxis = require("../../../core/components/axes/components/CategoryLeftAxis.js");
var import_ChartGridRoot = require("../../../core/components/chart-grid/ChartGridRoot.js");
var import_useCategoricalBarChartBaseConfig = require("../../hooks/useCategoricalBarChartBaseConfig.js");
var import_useCategoricalBarChartCategoryAxisConfig = require("../../hooks/useCategoricalBarChartCategoryAxisConfig.js");
var import_useCategoricalBarChartScales = require("../../hooks/useCategoricalBarChartScales.js");
var import_useCategoricalBarFilteredData = require("../../hooks/useCategoricalBarFilteredData.js");
var import_useTruncationMode = require("../../hooks/useTruncationMode.js");
var import_categorical_bar_area_rules = require("../../types/categorical-bar-area-rules.js");
var import_layout = require("../../utils/layout.js");
const CategoricalBarChartCategoryAxis = () => {
  const { filteredCategories, categoryMap } = (0, import_useCategoricalBarFilteredData.useCategoricalBarFilteredData)();
  const { categoryScale } = (0, import_useCategoricalBarChartScales.useCategoricalBarChartScales)();
  const { layout } = (0, import_useCategoricalBarChartBaseConfig.useCategoricalBarChartBaseConfig)();
  const truncationMode = (0, import_useTruncationMode.useTruncationMode)();
  const { label: categoryAxisLabel, tickLabelLayout } = (0, import_useCategoricalBarChartCategoryAxisConfig.useCategoricalBarChartCategoryAxisConfig)();
  const categoryTicks = (0, import_react.useMemo)(
    () => (0, import_categorical_category_axis.getTickCategories)(filteredCategories, categoryMap),
    [categoryMap, filteredCategories]
  );
  return (0, import_layout.isVertical)(layout) ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ChartGridRoot.ChartGrid.GridArea, { gridArea: import_categorical_bar_area_rules.AREAS.bottomAxis, children: ({ availableWidth, availableHeight }) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_CategoryBottomAxis.CategoryBottomAxis,
    {
      isVerticalLayout: true,
      categories: categoryTicks,
      containerHeight: availableHeight,
      containerWidth: availableWidth,
      scale: categoryScale,
      label: categoryAxisLabel,
      tickLabelLayout,
      truncationMode
    }
  ) }) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ChartGridRoot.ChartGrid.GridArea, { gridArea: import_categorical_bar_area_rules.AREAS.leftAxis, children: ({ availableWidth, availableHeight }) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_CategoryLeftAxis.CategoryLeftAxis,
    {
      isVerticalLayout: false,
      categories: categoryTicks,
      containerHeight: availableHeight,
      containerWidth: availableWidth,
      scale: categoryScale,
      label: categoryAxisLabel,
      truncationMode
    }
  ) });
};
CategoricalBarChartCategoryAxis["displayName"] = "CategoricalBarChartAxis";
