// @dynatrace/strato-components/content/information-overlay/InformationOverlay.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var InformationOverlay_exports = {};
__export(InformationOverlay_exports, {
  InformationOverlay: () => InformationOverlay
});
module.exports = __toCommonJS(InformationOverlay_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_Content = require("./Content.js");
var import_InformationOverlayContext = require("./contexts/InformationOverlayContext.js");
var import_information_overlay_constants = require("./information-overlay-constants.js");
var import_InformationOverlayIcon = require("./InformationOverlayIcon.js");
var import_Trigger = require("./Trigger.js");
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useControllableState = require("../../core/hooks/useControllableState.js");
var import_useId = require("../../core/hooks/useId.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_useSafeForwardProps = require("../../core/hooks/useSafeForwardProps.js");
var import_useSlots = require("../../core/hooks/useSlots.js");
var import_Flex = require("../../layouts/flex/Flex.js");
var import_useBreakpoint = require("../../layouts/hooks/useBreakpoint.js");
var import_useOverlayWithTrigger = require("../../overlays/hooks/useOverlayWithTrigger.js");
var import_Overlay = require("../../overlays/overlay/Overlay.js");
const InformationOverlayRoot = (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    children,
    defaultOpen,
    open,
    onOpenChange,
    color = "neutral",
    placement = "bottom",
    style: consumerStyle,
    className: consumerClassName,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const ariaLabelingProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(restProps);
  const forwardProps = (0, import_useSafeForwardProps.useSafeForwardProps)(restProps, ariaLabelingProps);
  const overlayContentId = (0, import_useId.useId)();
  const [isOpen, setIsOpen] = (0, import_useControllableState.useControllableState)({
    defaultValue: defaultOpen,
    value: open,
    onChange: onOpenChange
  });
  const matchesMaxWidthWindow = (0, import_useBreakpoint.useBreakpoint)("(max-width: 480px)");
  const {
    targetRef,
    triggerRef,
    overlayRef,
    overlayState,
    overlayProps,
    overlayTriggerProps
  } = (0, import_useOverlayWithTrigger.useOverlayWithTrigger)({
    placement: `${placement}-start`,
    placementFallbacks: [
      "top-start",
      "bottom-start",
      "right-start",
      "left-start"
    ],
    offset: 4,
    crossOffset: -16,
    widthStrategy: matchesMaxWidthWindow ? { max: 300 } : { max: 400 },
    open: isOpen,
    onOpenChange: setIsOpen,
    shouldCloseOnBlur: true
  });
  const overlayStateIsOpen = overlayState.isOpen;
  const rootRef = (0, import_useMergeRefs.useMergeRefs)([forwardedRef, triggerRef, targetRef]);
  const [slots] = (0, import_useSlots.useSlots)(children, {
    trigger: import_Trigger.Trigger,
    content: import_Content.Content
  });
  const { children: triggerChildren, ...triggerProps } = slots.trigger?.props ?? {};
  const contextProviderValue = (0, import_react.useMemo)(() => {
    return { color, overlayTriggerProps, id: overlayContentId, isOpen };
  }, [color, isOpen, overlayContentId, overlayTriggerProps]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_InformationOverlayContext.InformationOverlayContext.Provider, { value: contextProviderValue, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Trigger.Trigger,
      {
        ref: rootRef,
        id: triggerProps.id,
        className: triggerProps.className,
        style: triggerProps.style,
        "aria-label": triggerProps["aria-label"],
        "aria-describedby": overlayStateIsOpen ? overlayContentId : void 0,
        "data-testid": triggerProps["data-testid"],
        "data-dtrum-mask": triggerProps["data-dtrum-mask"] ?? dataDtrumMask,
        "data-dtrum-allow": triggerProps["data-dtrum-allow"] ?? dataDtrumAllow,
        ...triggerProps,
        children: triggerChildren
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Overlay.Overlay,
      {
        ref: overlayRef,
        isOpen: overlayStateIsOpen,
        overlayProps,
        className: consumerClassName,
        style: consumerStyle,
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        ...ariaLabelingProps,
        ...behavioralTrackingProps,
        ...forwardProps,
        "data-dt-component": import_information_overlay_constants.COMPONENT_NAME,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_Flex.Flex,
          {
            id: overlayContentId,
            paddingY: 12,
            paddingX: 16,
            flexDirection: "column",
            children: slots.content
          }
        )
      }
    )
  ] });
});
InformationOverlayRoot.displayName = import_information_overlay_constants.COMPONENT_NAME;
const InformationOverlay = Object.assign(InformationOverlayRoot, {
  Content: import_Content.Content,
  Trigger: import_Trigger.Trigger,
  Icon: import_InformationOverlayIcon.InformationOverlayIcon
});
