// @dynatrace/strato-components/content/information-overlay/Trigger.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Trigger_exports = {};
__export(Trigger_exports, {
  Trigger: () => Trigger
});
module.exports = __toCommonJS(Trigger_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_useInformationOverlayContext = require("./hooks/useInformationOverlayContext.js");
var import_information_overlay_constants = require("./information-overlay-constants.js");
var import_InformationOverlay_css = require("./InformationOverlay.sty.js");
var import_InformationOverlayIcon = require("./InformationOverlayIcon.js");
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useId = require("../../core/hooks/useId.js");
var import_useSlots = require("../../core/hooks/useSlots.js");
var import_useFocusRing = require("../../core/styles/useFocusRing.js");
var import_can_receive_default_text_format = require("../../core/utils/can-receive-default-text-format.js");
var import_merge_props = require("../../core/utils/merge-props.js");
const Trigger = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    children,
    id: propId,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    ...remainingProps
  } = props;
  const { color, overlayTriggerProps } = (0, import_useInformationOverlayContext.useInformationOverlayContext)();
  const id = (0, import_useId.useId)();
  const { focusProps, focusClassName, isFocusVisible } = (0, import_useFocusRing.useFocusRing)({
    color: children ? "neutral" : color,
    isMinimal: true
  });
  const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const {
    "data-dt-name": dataDtName = "Trigger",
    ...remainingBehavioralTrackingProps
  } = behavioralTrackingProps;
  const { "aria-label": ariaLabelProp, ...remainingAriaLabelingProps } = (0, import_useAriaLabelingProps.useAriaLabelingProps)(restProps);
  const [slots, freeform] = (0, import_useSlots.useSlots)(children, {
    icon: import_InformationOverlayIcon.InformationOverlayIcon
  });
  const hasFreeform = freeform.filter(
    (child) => child !== "" && child !== void 0 && child !== null
  ).length > 0;
  if (slots.icon === void 0) {
    switch (color) {
      case "success":
        slots.icon = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.SuccessIcon, {});
        break;
      case "warning":
        slots.icon = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.WarningIcon, {});
        break;
      case "critical":
        slots.icon = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.CriticalIcon, {});
        break;
      default:
        slots.icon = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.InformationIcon, {});
        break;
    }
  }
  const intl = (0, import_react_intl.useIntl)();
  let ariaLabel = ariaLabelProp;
  if (ariaLabel === void 0 && (children === void 0 || !(0, import_can_receive_default_text_format.canReceiveDefaultTextFormat)(children))) {
    ariaLabel = intl.formatMessage({
      defaultMessage: "Open additional information.",
      id: "NXdoElYQZNKea1Sc",
      description: "Aria label fallback for Information overlay trigger"
    });
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    "button",
    {
      ref: forwardedRef,
      "aria-haspopup": "dialog",
      "aria-label": ariaLabel,
      id: propId ?? id,
      ...(0, import_merge_props.mergeProps)(
        overlayTriggerProps,
        focusProps,
        remainingAriaLabelingProps,
        remainingBehavioralTrackingProps
      ),
      className: (0, import_clsx.default)(import_InformationOverlay_css.InformationOverlayTriggerCSS, consumerClassName),
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "data-dt-name": dataDtName,
      "data-dt-component": import_information_overlay_constants.COMPONENT_NAME,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_InformationOverlayIcon.InformationOverlayIcon,
          {
            className: (0, import_clsx.default)(
              import_InformationOverlay_css.InformationOverlayTriggerIconCSS,
              import_InformationOverlay_css.InformationOverlayTriggerIconVariantsCSS[color],
              {
                [focusClassName]: !hasFreeform,
                [import_InformationOverlay_css.InformationOverlayTriggerIconHoverCSS]: !hasFreeform
              },
              { "strato-focus-visible": isFocusVisible }
            ),
            children: slots.icon
          }
        ),
        hasFreeform && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          "span",
          {
            className: (0, import_clsx.default)(focusClassName, import_InformationOverlay_css.InformationOverlayTriggerTextCSS, {
              "strato-focus-visible": isFocusVisible
            }),
            style: {
              flex: 1,
              overflow: "hidden",
              textDecoration: isFocusVisible ? "none" : void 0
            },
            children: freeform
          }
        )
      ]
    }
  );
});
Trigger.displayName = "InformationOverlay.Trigger";
