// @dynatrace/strato-components/charts/xy-chart/components/overlay/utils/chartCrosshairDataTransformer.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var chartCrosshairDataTransformer_exports = {};
__export(chartCrosshairDataTransformer_exports, {
  chartCrosshairDataTransformer: () => chartCrosshairDataTransformer
});
module.exports = __toCommonJS(chartCrosshairDataTransformer_exports);
var import_units = require("@dynatrace-sdk/units");
var import_get_axis_formatter_with_time_fallback = require("./get-axis-formatter-with-time-fallback.js");
var import_get_magnifier_label = require("../../../../core/components/crosshair/utils/get-magnifier-label.js");
var import_defaultUnitFormatter = require("../../../../core/utils/defaultUnitFormatter.js");
const coordinates = {
  top: "x",
  bottom: "x",
  left: "y",
  right: "y"
};
const defaultLabels = () => ({
  left: void 0,
  right: void 0,
  bottom: void 0,
  top: void 0
});
const accumulateFormatters = (getFormatter) => (acc, axis) => {
  const formatter = getFormatter(axis);
  if (formatter) {
    acc.push([axis, formatter]);
  }
  return acc;
};
const accumulateLabels = (getLabel) => (acc, axisWithFormatter) => {
  const [axis, formatter] = axisWithFormatter;
  const label = getLabel(axis, formatter);
  if (label) {
    acc[axis] = label;
  }
  return acc;
};
const chartCrosshairDataTransformer = ({
  axesScale,
  axesConfig,
  axesIdToFormatter,
  width,
  axes = ["left", "right", "bottom", "top"],
  formatCategory
}) => {
  const getFormatter = (axis) => {
    const axisConfig = axesConfig[axis];
    const scale = axesScale[axis];
    if (axisConfig) {
      if (axisConfig.isRelative) {
        return (0, import_defaultUnitFormatter.defaultUnitFormatter)(import_units.units.percentage.percent);
      }
      return (0, import_get_axis_formatter_with_time_fallback.getAxisFormatterWithTimeFallback)({
        axisConfig,
        axisId: axisConfig.id,
        scale,
        axesIdToFormatter,
        width
      });
    }
  };
  const formatters = axes.reduce(
    accumulateFormatters(getFormatter),
    []
  );
  return ({ position, pinned }) => {
    if (pinned) {
      return defaultLabels();
    }
    const getLabel = (axis, formatter) => {
      const raw = (0, import_get_magnifier_label.getMagnifierLabel)(
        axesConfig[axis]?.unit,
        coordinates[axis],
        formatter,
        position,
        axesScale[axis]
      );
      if (raw === null || !formatCategory) {
        return raw;
      }
      return formatCategory(axis, raw) ?? raw;
    };
    return formatters.reduce(accumulateLabels(getLabel), defaultLabels());
  };
};
