// @dynatrace/strato-components/charts/xy-chart/components/overlay/utils/get-selected-datapoints.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var get_selected_datapoints_exports = {};
__export(get_selected_datapoints_exports, {
  buildContentHighlightedPoint: () => buildContentHighlightedPoint,
  buildContentHighlightedRectangle: () => buildContentHighlightedRectangle,
  buildRectangleHighlightedPoints: () => buildRectangleHighlightedPoints,
  buildSelectionHighlightedPoints: () => buildSelectionHighlightedPoints,
  buildSelectionHighlightedRectangles: () => buildSelectionHighlightedRectangles,
  collectLinearXAxisIds: () => collectLinearXAxisIds,
  getSelectedDatapoints: () => getSelectedDatapoints,
  hasLinearContentPoint: () => hasLinearContentPoint
});
module.exports = __toCommonJS(get_selected_datapoints_exports);
var import_lodash_es = require("lodash");
var import_is_rectangle_selected = require("./is-rectangle-selected.js");
var import_serialize_linear_series_values = require("./serialize-linear-series-values.js");
var import_serialize_rectangle_series_values = require("./serialize-rectangle-series-values.js");
var import_constants = require("../../../../core/components/data-point/constants.js");
var import_get_selection_pixels = require("../../../../core/utils/get-selection-pixels.js");
var import_safe_get = require("../../../../core/utils/safe-get.js");
var import_scale = require("../../../../core/utils/scale.js");
var import_guards = require("../../../types/guards.js");
var import_xy_chart_internals = require("../../../types/xy-chart-internals.js");
var import_get_linear_accessors = require("../../../utils/get-linear-accessors.js");
var import_get_scale = require("../../../utils/get-scale.js");
var import_type_guards = require("../../../utils/type-guards.js");
var import_validation_datapoint = require("../../../utils/validation-datapoint.js");
var import_value_representation = require("../../../utils/value-representation.js");
var import_area_accessors = require("../../displays/area-accessors.js");
var import_band_accessors = require("../../displays/band-accessors.js");
var import_bar_accessors = require("../../displays/bar-accessors.js");
var import_BarDisplay = require("../../displays/BarDisplay/BarDisplay.js");
var import_dot_accessors = require("../../displays/dot-accessors.js");
var import_line_accessors = require("../../displays/line-accessors.js");
var import_rect_accessors = require("../../displays/rect-accessors.js");
function getCategoriesIncluded(categories, domainStart, domainEnd) {
  const startIndex = categories.indexOf(domainStart);
  const endIndex = categories.indexOf(domainEnd);
  return categories.slice(startIndex, endIndex);
}
function getSelectionDomain(selection, axisConfig) {
  if ((0, import_guards.isAnyNumericalScale)(axisConfig)) {
    return [
      axisConfig.axisScale.invert(selection.start),
      axisConfig.axisScale.invert(selection.end)
    ];
  } else if ((0, import_guards.isCategoricalScale)(axisConfig)) {
    const domainStart = (0, import_scale.scaleBandInvert)(axisConfig.axisScale)(selection.start);
    const domainEnd = (0, import_scale.scaleBandInvert)(axisConfig.axisScale)(selection.end);
    const categories = axisConfig.axisScale.domain();
    return getCategoriesIncluded(categories, domainStart, domainEnd);
  }
  return [selection.start, selection.end];
}
const getSelectedDatapoints = (seriesWithDatapoints, axisIdToScales, axes, selection, dotSeriesToMetadata, lineSeriesToMetadata, rectSeriesToMetadata, datapointMetadataMap, axisIdToScaleDomain, barSeriesToMetadata, areaSeriesToMetadata) => {
  const rectSeriesIndex = { current: 0 };
  const isDefined = (0, import_validation_datapoint.isGapDatapointForDisplay)(datapointMetadataMap);
  return seriesWithDatapoints.map((series) => {
    const xScale = (0, import_get_scale.getCurrentScale)(axisIdToScales, series.xAxisId);
    const yScale = (0, import_get_scale.getCurrentScale)(axisIdToScales, series.yAxisId);
    const [yMin, yMax] = yScale.range().toSorted((a, b) => a - b);
    const axisConfig = {
      axisType: axes.find((axis) => axis.id === series.xAxisId)?.type,
      axisScale: xScale
    };
    if ((0, import_type_guards.isDotSeries)(series)) {
      const { xAxis, yAxis } = (0, import_safe_get.safeGet)(dotSeriesToMetadata, series);
      const xRelAccessor = (0, import_value_representation.relAccessor)(xAxis, axisIdToScaleDomain);
      const yRelAccessor = (0, import_value_representation.relAccessor)(yAxis, axisIdToScaleDomain);
      return getDotSeriesSelection(
        series,
        dotSeriesToMetadata,
        xScale,
        yScale,
        selection,
        [yMin, yMax],
        axisConfig,
        xRelAccessor,
        yRelAccessor
      );
    } else if ((0, import_type_guards.isLineSeries)(series)) {
      const { xAxis, yAxis } = (0, import_safe_get.safeGet)(lineSeriesToMetadata, series);
      const xRelAccessor = (0, import_value_representation.relAccessor)(xAxis, axisIdToScaleDomain);
      const yRelAccessor = (0, import_value_representation.relAccessor)(yAxis, axisIdToScaleDomain);
      return getLineSeriesSelection(
        series,
        lineSeriesToMetadata,
        xScale,
        yScale,
        selection,
        [yMin, yMax],
        axisConfig,
        isDefined,
        xRelAccessor,
        yRelAccessor
      );
    } else if ((0, import_type_guards.isRectSeries)(series)) {
      const { xAxis, yAxis } = (0, import_safe_get.safeGet)(rectSeriesToMetadata, series);
      const xRelAccessor = (0, import_value_representation.relAccessor)(xAxis, axisIdToScaleDomain);
      const yRelAccessor = (0, import_value_representation.relAccessor)(yAxis, axisIdToScaleDomain);
      return getRectSeriesSelection(series, {
        rectSeriesIndex,
        xScale,
        yScale,
        selection,
        yRange: [yMin, yMax],
        axisConfig,
        isDefined,
        xRelAccessor,
        yRelAccessor
      });
    } else if ((0, import_type_guards.isBarSeries)(series)) {
      const { xAxis, yAxis } = (0, import_safe_get.safeGet)(barSeriesToMetadata, series);
      const xRelAccessor = (0, import_value_representation.relAccessor)(xAxis, axisIdToScaleDomain);
      const yRelAccessor = (0, import_value_representation.relAccessor)(yAxis, axisIdToScaleDomain);
      return getBarSeriesSelection(
        barSeriesToMetadata,
        series,
        selection,
        [yMin, yMax],
        axisConfig,
        datapointMetadataMap,
        axisIdToScales,
        xRelAccessor,
        yRelAccessor
      );
    } else if ((0, import_type_guards.isAreaSeries)(series)) {
      const { xAxis, yAxis } = (0, import_safe_get.safeGet)(areaSeriesToMetadata, series);
      const xRelAccessor = (0, import_value_representation.relAccessor)(xAxis, axisIdToScaleDomain);
      const yRelAccessor = (0, import_value_representation.relAccessor)(yAxis, axisIdToScaleDomain);
      return getAreaSeriesSelection(series, {
        areaSeriesToMetadata,
        xScale,
        yScale,
        selection,
        yRange: [yMin, yMax],
        axisConfig,
        isDefined,
        datapointMetadataMap,
        xRelAccessor,
        yRelAccessor
      });
    } else {
      return {
        seriesName: "",
        datapoints: [],
        domain: getSelectionDomain(selection, axisConfig)
      };
    }
  });
};
function getDotSeriesSelection(series, dotSeriesToMetadata, xScale, yScale, selection, yRange, axisConfig, xRelAccessor, yRelAccessor) {
  const filteredData = filterPointSeries(
    series.data,
    (0, import_dot_accessors.xAccessor)(xScale, xRelAccessor),
    (0, import_dot_accessors.yAccessor)(yScale, yRelAccessor),
    selection,
    yRange
  );
  return {
    seriesName: (0, import_safe_get.safeGet)(dotSeriesToMetadata, series).name,
    datapoints: filteredData,
    domain: getSelectionDomain(selection, axisConfig)
  };
}
function getLineSeriesSelection(series, lineSeriesToMetadata, xScale, yScale, selection, yRange, axisConfig, isDefined, xRelAccessor, yRelAccessor) {
  const filteredData = filterPointSeries(
    series.data,
    (0, import_line_accessors.xAccessor)(xScale, xRelAccessor),
    (0, import_line_accessors.yAccessor)(yScale, yRelAccessor),
    selection,
    yRange,
    isDefined
  );
  return {
    seriesName: (0, import_safe_get.safeGet)(lineSeriesToMetadata, series).name,
    datapoints: filteredData,
    domain: getSelectionDomain(selection, axisConfig)
  };
}
function isWithinVerticalDomain(yStart, yEnd, yMin, yMax) {
  const top = Math.min(yStart, yEnd);
  const bottom = Math.max(yStart, yEnd);
  return top >= yMin && bottom <= yMax;
}
function getRectSeriesSelection(series, ctx) {
  const {
    rectSeriesIndex,
    xScale,
    yScale,
    selection,
    yRange,
    axisConfig,
    isDefined,
    xRelAccessor,
    yRelAccessor
  } = ctx;
  const xAccessor2 = (0, import_rect_accessors.xAccessor)(xScale, xRelAccessor);
  const yAccessor2 = (0, import_rect_accessors.yAccessor)(yScale, yRelAccessor);
  const horizontalBandAccessor = (0, import_rect_accessors.widthAccessor)(xScale, xRelAccessor);
  const verticalBandAccessor = (0, import_rect_accessors.heightAccessor)(yScale, yRelAccessor);
  const [yMin, yMax] = yRange;
  const filteredDatapoints = series.filteredData.filter((datapoint) => {
    if (!isDefined(datapoint)) {
      return false;
    }
    const rectXStart = xAccessor2(datapoint);
    const rectXEnd = rectXStart + horizontalBandAccessor(datapoint);
    const rectYStart = yAccessor2(datapoint);
    const rectYEnd = rectYStart + verticalBandAccessor(datapoint);
    if (!isWithinVerticalDomain(rectYStart, rectYEnd, yMin, yMax)) {
      return false;
    }
    return (0, import_is_rectangle_selected.isRectangleSelected)({
      x: {
        rectangleStart: rectXStart,
        rectangleEnd: rectXEnd,
        selectionStart: selection.start,
        selectionEnd: selection.end
      },
      y: {
        rectangleStart: rectYStart,
        rectangleEnd: rectYEnd,
        selectionStart: yMin,
        selectionEnd: yMax
      }
    });
  });
  return {
    seriesName: `rect-series-${++rectSeriesIndex.current}`,
    datapoints: filteredDatapoints,
    domain: getSelectionDomain(selection, axisConfig)
  };
}
function getBarSeriesSelection(barSeriesToMetadata, series, selection, yRange, axisConfig, datapointMetadataMap, axisIdToScales, xRelAccessor, yRelAccessor) {
  const xScale = (0, import_get_scale.getCurrentScale)(
    axisIdToScales,
    series.xAxisId,
    (0, import_BarDisplay.isXBaselineAccessor)(series.data[0]) ? series.type : void 0
  );
  const yScale = (0, import_get_scale.getCurrentScale)(
    axisIdToScales,
    series.yAxisId,
    (0, import_BarDisplay.isYBaselineAccessor)(series.data[0]) ? series.type : void 0
  );
  const [yMin, yMax] = yRange;
  const dimensionXScale = (0, import_get_scale.getDimensionScale)(axisIdToScales, series.xAxisId);
  const dimensionYScale = (0, import_get_scale.getDimensionScale)(axisIdToScales, series.yAxisId);
  const xAccessor2 = (0, import_bar_accessors.xAccessor)(
    xScale,
    datapointMetadataMap,
    dimensionXScale,
    series.id,
    xRelAccessor
  );
  const yAccessor2 = (0, import_bar_accessors.yAccessor)(
    yScale,
    datapointMetadataMap,
    dimensionYScale,
    series.id,
    yRelAccessor
  );
  const horizontalBandAccessor = (0, import_bar_accessors.widthAccessor)(
    xScale,
    datapointMetadataMap,
    dimensionXScale,
    xRelAccessor
  );
  const verticalBandAccessor = (0, import_bar_accessors.heightAccessor)(
    yScale,
    datapointMetadataMap,
    dimensionYScale,
    yRelAccessor
  );
  const filteredData = series.data.filter((datapoint) => {
    const barXStart = xAccessor2(datapoint);
    const barXEnd = barXStart + horizontalBandAccessor(datapoint);
    const barYStart = yAccessor2(datapoint);
    const barYEnd = barYStart + verticalBandAccessor(datapoint);
    if (!isWithinVerticalDomain(barYStart, barYEnd, yMin, yMax)) {
      return false;
    }
    return (0, import_is_rectangle_selected.isRectangleSelected)({
      x: {
        rectangleStart: barXStart,
        rectangleEnd: barXEnd,
        selectionStart: selection.start,
        selectionEnd: selection.end
      },
      y: {
        rectangleStart: barYStart,
        rectangleEnd: barYEnd,
        selectionStart: yMin,
        selectionEnd: yMax
      }
    });
  });
  return {
    seriesName: (0, import_safe_get.safeGet)(barSeriesToMetadata, series).name,
    datapoints: filteredData,
    domain: getSelectionDomain(selection, axisConfig)
  };
}
function getAreaSeriesSelection(series, ctx) {
  const {
    areaSeriesToMetadata,
    xScale,
    yScale,
    selection,
    yRange,
    axisConfig,
    isDefined,
    datapointMetadataMap,
    xRelAccessor,
    yRelAccessor
  } = ctx;
  const { name } = (0, import_safe_get.safeGet)(areaSeriesToMetadata, series);
  const [yMin, yMax] = yRange;
  const { xAccessor: xAreaAccessor, yAccessor: yAreaAccessor } = (0, import_get_linear_accessors.getLinearAccessors)(
    import_xy_chart_internals.AREA_VARIANT,
    (0, import_safe_get.safeGet)(areaSeriesToMetadata, series),
    datapointMetadataMap,
    xRelAccessor,
    yRelAccessor
  );
  const filteredData = series.data.filter((datapoint) => {
    if (!isDefined(datapoint)) {
      return false;
    }
    const xCoord = xAreaAccessor(xScale)(datapoint);
    const yCoord = yAreaAccessor(yScale)(datapoint);
    return yCoord >= yMin && yCoord <= yMax && xCoord >= selection.start && xCoord <= selection.end;
  });
  return {
    seriesName: name,
    datapoints: filteredData,
    domain: getSelectionDomain(selection, axisConfig)
  };
}
function filterPointSeries(data, xAccessor2, yAccessor2, selection, yRange, isDefined) {
  const [yMin, yMax] = yRange;
  return data.filter((datapoint) => {
    if (isDefined && !isDefined(datapoint)) {
      return false;
    }
    const pointX = xAccessor2(datapoint);
    const pointY = yAccessor2(datapoint);
    return pointX >= selection.start && pointX <= selection.end && pointY >= yMin && pointY <= yMax;
  });
}
function hasLinearContentPoint(contentState) {
  return !!contentState?.series && !!contentState?.datapoint && (contentState.type === import_xy_chart_internals.DOT_VARIANT || contentState.type === import_xy_chart_internals.LINE_VARIANT || contentState.type === import_xy_chart_internals.AREA_VARIANT);
}
function collectLinearXAxisIds(dotSeries, lineSeries, areaSeries) {
  const ids = /* @__PURE__ */ new Set();
  for (const series of dotSeries) {
    if (series.xAxisId) {
      ids.add(series.xAxisId);
    }
  }
  for (const series of lineSeries) {
    if (series.xAxisId) {
      ids.add(series.xAxisId);
    }
  }
  for (const series of areaSeries) {
    if (series.xAxisId) {
      ids.add(series.xAxisId);
    }
  }
  return ids;
}
function getAxisIdToIsRelative(axes) {
  const axisIdToIsRelative = /* @__PURE__ */ new Map();
  axes.forEach((axis) => {
    axisIdToIsRelative.set(axis.id, axis.isRelative);
  });
  return axisIdToIsRelative;
}
function buildSelectionHighlightedPoints(params) {
  const {
    selectState,
    linearSelection,
    width,
    axisIdToScales,
    dotSeriesWithDatapoints,
    lineSeriesWithDatapoints,
    areaSeriesWithDatapoints,
    dotSeriesMeta,
    lineSeriesMeta,
    areaSeriesMeta,
    axisIdToScaleDomain,
    axes,
    datapointMetadataMap
  } = params;
  const anySelection = !selectState || linearSelection.selectionConfigByAxisId.size === 0;
  if (anySelection) {
    return [];
  }
  const axisIds = collectLinearXAxisIds(
    dotSeriesWithDatapoints,
    lineSeriesWithDatapoints,
    areaSeriesWithDatapoints
  );
  const axisIdToIsRelative = getAxisIdToIsRelative(axes);
  if (!axisIds.size) {
    return [];
  }
  const [selectionStart, selectionEnd] = (0, import_get_selection_pixels.getSelectionPixels)(
    selectState.selectionPercentage,
    width,
    [selectState.start, selectState.end]
  );
  const { highlightedDatapoints } = linearSelection.searchAllAxisDatapoints(
    selectionStart.relative.x,
    selectionEnd.relative.x,
    (0, import_serialize_linear_series_values.serializeLinearSeriesValues)(
      (series) => getLinearSeriesMetadata(
        series,
        dotSeriesMeta,
        lineSeriesMeta,
        areaSeriesMeta
      ),
      axisIdToScales,
      axisIdToScaleDomain,
      datapointMetadataMap
    ),
    Array.from(axisIds),
    axisIdToScales,
    axisIdToScaleDomain,
    axisIdToIsRelative
  );
  return highlightedDatapoints;
}
function buildDotHighlightedPoint(datapoint, meta, axisIdToScales, axisIdToScaleDomain) {
  const xScale = (0, import_get_scale.getCurrentScale)(axisIdToScales, meta.xAxis.id);
  const yScale = (0, import_get_scale.getCurrentScale)(axisIdToScales, meta.yAxis.id);
  const { xAxis, yAxis } = meta;
  const xRelAccessor = (0, import_value_representation.relAccessor)(xAxis, axisIdToScaleDomain);
  const yRelAccessor = (0, import_value_representation.relAccessor)(yAxis, axisIdToScaleDomain);
  return [
    {
      id: import_xy_chart_internals.DOT_ID,
      x: (0, import_dot_accessors.xAccessor)(xScale, xRelAccessor)(datapoint) + import_constants.POINT_OVERLAY_PADDING,
      y: (0, import_dot_accessors.yAccessor)(yScale, yRelAccessor)(datapoint) + import_constants.POINT_OVERLAY_PADDING,
      color: meta.color,
      shape: meta.shape
    }
  ];
}
function buildLineHighlightedPoint(datapoint, meta, axisIdToScales, axisIdToScaleDomain) {
  const xScale = (0, import_get_scale.getCurrentScale)(axisIdToScales, meta.xAxis.id);
  const yScale = (0, import_get_scale.getCurrentScale)(axisIdToScales, meta.yAxis.id);
  const { xAxis, yAxis } = meta;
  const xRelAccessor = (0, import_value_representation.relAccessor)(xAxis, axisIdToScaleDomain);
  const yRelAccessor = (0, import_value_representation.relAccessor)(yAxis, axisIdToScaleDomain);
  return [
    {
      id: import_xy_chart_internals.LINE_ID,
      x: (0, import_line_accessors.xAccessor)(xScale, xRelAccessor)(datapoint) + import_constants.POINT_OVERLAY_PADDING,
      y: (0, import_line_accessors.yAccessor)(yScale, yRelAccessor)(datapoint) + import_constants.POINT_OVERLAY_PADDING,
      color: meta.color,
      shape: void 0
    }
  ];
}
function buildHorizontalAreaPoints(datapoint, yScale, xScale, isBand, color, datapointMetadataMap, xRelAccessor, yRelAccessor, isXStacked) {
  const yCoord = isBand ? (0, import_band_accessors.bandY0Accessor)(yScale, yRelAccessor)(datapoint) + import_constants.POINT_OVERLAY_PADDING : (0, import_area_accessors.yAccessor)(yScale, yRelAccessor)(datapoint) + import_constants.POINT_OVERLAY_PADDING;
  const points = [];
  if (isBand) {
    points.push({
      id: import_xy_chart_internals.AREA_ID,
      x: (0, import_band_accessors.bandX0Accessor)(xScale, xRelAccessor)(datapoint) + import_constants.POINT_OVERLAY_PADDING,
      y: yCoord,
      color,
      shape: void 0
    });
  }
  points.push({
    id: import_xy_chart_internals.AREA_ID,
    x: (isBand ? (0, import_band_accessors.bandX1Accessor)(xScale, xRelAccessor) : (0, import_area_accessors.x1Accessor)(xScale, datapointMetadataMap, xRelAccessor, isXStacked))(
      datapoint
    ) + import_constants.POINT_OVERLAY_PADDING,
    y: yCoord,
    color,
    shape: void 0
  });
  return points;
}
function buildVerticalAreaPoints(datapoint, xScale, yScale, isBand, color, datapointMetadataMap, xRelAccessor, yRelAccessor, isYStacked) {
  const xCoord = isBand ? (0, import_band_accessors.bandX0Accessor)(xScale, xRelAccessor)(datapoint) + import_constants.POINT_OVERLAY_PADDING : (0, import_area_accessors.xAccessor)(xScale, xRelAccessor)(datapoint) + import_constants.POINT_OVERLAY_PADDING;
  const points = [];
  if (isBand) {
    points.push({
      id: import_xy_chart_internals.AREA_ID,
      x: xCoord,
      y: (0, import_band_accessors.bandY0Accessor)(yScale, yRelAccessor)(datapoint) + import_constants.POINT_OVERLAY_PADDING,
      color,
      shape: void 0
    });
  }
  points.push({
    id: import_xy_chart_internals.AREA_ID,
    x: xCoord,
    y: (isBand ? (0, import_band_accessors.bandY1Accessor)(yScale, yRelAccessor) : (0, import_area_accessors.y1Accessor)(yScale, datapointMetadataMap, yRelAccessor, isYStacked))(
      datapoint
    ) + import_constants.POINT_OVERLAY_PADDING,
    color,
    shape: void 0
  });
  return points;
}
function buildAreaHighlightedPoints(datapoint, meta, axisIdToScales, datapointMetadataMap, axisIdToScaleDomain) {
  const isHorizontal = meta.orientation === "horizontal";
  const isBand = meta.variant === "band";
  const xScale = (0, import_get_scale.getCurrentScale)(axisIdToScales, meta.xAxis.id);
  const yScale = (0, import_get_scale.getCurrentScale)(axisIdToScales, meta.yAxis.id);
  if (!xScale || !yScale) {
    return [];
  }
  const { xAxis, yAxis, isStacked } = meta;
  const xRelAccessor = (0, import_value_representation.relAccessor)(xAxis, axisIdToScaleDomain);
  const yRelAccessor = (0, import_value_representation.relAccessor)(yAxis, axisIdToScaleDomain);
  if (isHorizontal) {
    return buildHorizontalAreaPoints(
      datapoint,
      yScale,
      xScale,
      isBand,
      meta.color,
      datapointMetadataMap,
      xRelAccessor,
      yRelAccessor,
      isStacked
    );
  } else {
    return buildVerticalAreaPoints(
      datapoint,
      xScale,
      yScale,
      isBand,
      meta.color,
      datapointMetadataMap,
      xRelAccessor,
      yRelAccessor,
      isStacked
    );
  }
}
function buildContentHighlightedPoint(params) {
  const {
    contentState,
    axisIdToScales,
    dotSeriesMeta,
    lineSeriesMeta,
    areaSeriesMeta,
    axisIdToScaleDomain,
    datapointMetadataMap
  } = params;
  if (!hasLinearContentPoint(contentState)) {
    return [];
  }
  if ((0, import_guards.isDotContentPoint)(contentState)) {
    const { series, datapoint } = contentState;
    const meta = (0, import_safe_get.safeGet)(dotSeriesMeta, series);
    return buildDotHighlightedPoint(
      datapoint,
      meta,
      axisIdToScales,
      axisIdToScaleDomain
    );
  }
  if ((0, import_guards.isLineContentPoint)(contentState)) {
    const { series, datapoint } = contentState;
    const meta = (0, import_safe_get.safeGet)(lineSeriesMeta, series);
    return buildLineHighlightedPoint(
      datapoint,
      meta,
      axisIdToScales,
      axisIdToScaleDomain
    );
  }
  if ((0, import_guards.isAreaContentPoint)(contentState)) {
    const { series, datapoint } = contentState;
    const meta = (0, import_safe_get.safeGet)(areaSeriesMeta, series);
    return buildAreaHighlightedPoints(
      datapoint,
      meta,
      axisIdToScales,
      datapointMetadataMap,
      axisIdToScaleDomain
    );
  }
  return [];
}
function getLinearSeriesMetadata(series, dotSeriesMeta, lineSeriesMeta, areaSeriesMeta) {
  if (series.type === import_xy_chart_internals.DOT_VARIANT) {
    return (0, import_safe_get.safeGet)(dotSeriesMeta, series);
  }
  if (series.type === import_xy_chart_internals.LINE_VARIANT) {
    return (0, import_safe_get.safeGet)(lineSeriesMeta, series);
  }
  return (0, import_safe_get.safeGet)(areaSeriesMeta, series);
}
function buildSelectionHighlightedRectangles(params) {
  const {
    selectState,
    rectSelection,
    width,
    rectSeriesMeta,
    rectSeriesWithDatapoints,
    barSeriesWithDatapoints,
    barSeriesMeta,
    axisIdToScales,
    axisIdToScaleDomain,
    axes,
    datapointMetadataMap
  } = params;
  if (!selectState || rectSelection.selectionConfigByAxisId.size === 0) {
    return [];
  }
  const definedAxes = Array.from(
    new Set(
      [...rectSeriesWithDatapoints, ...barSeriesWithDatapoints].flatMap(
        (series) => {
          const axes2 = [];
          if (!(0, import_lodash_es.isUndefined)(series.xAxisId)) {
            axes2.push(series.xAxisId);
          }
          if (!(0, import_lodash_es.isUndefined)(series.yAxisId)) {
            axes2.push(series.yAxisId);
          }
          return axes2;
        }
      )
    )
  );
  const axisIdToIsRelative = getAxisIdToIsRelative(axes);
  if (!definedAxes.length) {
    return [];
  }
  const [selectionStart, selectionEnd] = (0, import_get_selection_pixels.getSelectionPixels)(
    selectState.selectionPercentage,
    width,
    [selectState.start, selectState.end]
  );
  const { highlightedDatapoints } = rectSelection.searchAllAxisDatapoints(
    selectionStart.relative.x,
    selectionEnd.relative.x,
    (0, import_serialize_rectangle_series_values.serializeRectangleSeriesValues)(
      axisIdToScales,
      rectSeriesMeta,
      axisIdToScaleDomain,
      datapointMetadataMap,
      barSeriesMeta
    ),
    definedAxes,
    axisIdToScales,
    axisIdToScaleDomain,
    axisIdToIsRelative
  );
  return highlightedDatapoints;
}
function buildContentHighlightedRectangle(params) {
  const {
    contentState,
    axisIdToScales,
    barSeriesMeta,
    rectSeriesMeta,
    datapointMetadataMap,
    axisIdToScaleDomain
  } = params;
  if (!contentState?.series || !contentState?.datapoint || contentState.type !== import_xy_chart_internals.BAR_VARIANT && contentState.type !== import_xy_chart_internals.RECT_VARIANT) {
    return null;
  }
  if (contentState.type === import_xy_chart_internals.BAR_VARIANT) {
    const { datapoint: datapoint2, series: series2 } = contentState;
    const { xAxis: xAxis2, yAxis: yAxis2 } = (0, import_safe_get.safeGet)(barSeriesMeta, series2);
    const xScale2 = (0, import_get_scale.getCurrentScale)(
      axisIdToScales,
      xAxis2.id,
      (0, import_BarDisplay.isXBaselineAccessor)(series2.data[0]) ? series2.type : void 0
    );
    const dimensionXScale = (0, import_get_scale.getDimensionScale)(axisIdToScales, series2.xAxisId);
    const yScale2 = (0, import_get_scale.getCurrentScale)(
      axisIdToScales,
      yAxis2.id,
      (0, import_BarDisplay.isYBaselineAccessor)(series2.data[0]) ? series2.type : void 0
    );
    const dimensionYScale = (0, import_get_scale.getDimensionScale)(axisIdToScales, series2.yAxisId);
    const xRelAccessor2 = (0, import_value_representation.relAccessor)(xAxis2, axisIdToScaleDomain);
    const yRelAccessor2 = (0, import_value_representation.relAccessor)(yAxis2, axisIdToScaleDomain);
    return {
      id: import_xy_chart_internals.BAR_ID,
      x: (0, import_bar_accessors.xAccessor)(
        xScale2,
        datapointMetadataMap,
        dimensionXScale,
        series2.id,
        xRelAccessor2
      )(datapoint2),
      y: (0, import_bar_accessors.yAccessor)(
        yScale2,
        datapointMetadataMap,
        dimensionYScale,
        series2.id,
        yRelAccessor2
      )(datapoint2),
      width: (0, import_bar_accessors.widthAccessor)(
        xScale2,
        datapointMetadataMap,
        dimensionXScale,
        xRelAccessor2
      )(datapoint2),
      height: (0, import_bar_accessors.heightAccessor)(
        yScale2,
        datapointMetadataMap,
        dimensionYScale,
        yRelAccessor2
      )(datapoint2)
    };
  }
  const { datapoint, series } = contentState;
  const { xAxis, yAxis } = (0, import_safe_get.safeGet)(rectSeriesMeta, series);
  const xScale = (0, import_get_scale.getCurrentScale)(axisIdToScales, xAxis.id);
  const yScale = (0, import_get_scale.getCurrentScale)(axisIdToScales, yAxis.id);
  const xRelAccessor = (0, import_value_representation.relAccessor)(xAxis, axisIdToScaleDomain);
  const yRelAccessor = (0, import_value_representation.relAccessor)(yAxis, axisIdToScaleDomain);
  return {
    id: import_xy_chart_internals.RECT_ID,
    x: (0, import_rect_accessors.xAccessor)(xScale, xRelAccessor)(datapoint),
    y: (0, import_rect_accessors.yAccessor)(yScale, yRelAccessor)(datapoint),
    width: (0, import_rect_accessors.widthAccessor)(xScale, xRelAccessor)(datapoint),
    height: (0, import_rect_accessors.heightAccessor)(yScale, yRelAccessor)(datapoint)
  };
}
function buildRectangleHighlightedPoints(params) {
  const selectionRects = buildSelectionHighlightedRectangles(params);
  const contentRect = buildContentHighlightedRectangle(params);
  return contentRect ? [...selectionRects, contentRect] : selectionRects;
}
