// @dynatrace/strato-components/charts/xy-chart/components/overlay/utils/serialize-linear-series-values.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var serialize_linear_series_values_exports = {};
__export(serialize_linear_series_values_exports, {
  serializeLinearSeriesValues: () => serializeLinearSeriesValues
});
module.exports = __toCommonJS(serialize_linear_series_values_exports);
var import_constants = require("../../../../core/components/data-point/constants.js");
var import_safe_get = require("../../../../core/utils/safe-get.js");
var import_xy_chart_internals = require("../../../types/xy-chart-internals.js");
var import_get_linear_accessors = require("../../../utils/get-linear-accessors.js");
var import_get_scale = require("../../../utils/get-scale.js");
var import_value_representation = require("../../../utils/value-representation.js");
var import_band_accessors = require("../../displays/band-accessors.js");
function getYRangeMinMax(range) {
  let min = range[0];
  let max = range[0];
  for (let i = 1; i < range.length; i++) {
    const v = range[i];
    if (v < min) {
      min = v;
    } else if (v > max) {
      max = v;
    }
  }
  return min < max ? [min, max] : [max, min];
}
function resolveMetadata(getter, series) {
  if (typeof getter === "function") {
    return getter(series);
  }
  return (0, import_safe_get.safeGet)(getter, series);
}
const idMap = {
  [import_xy_chart_internals.DOT_VARIANT]: import_xy_chart_internals.DOT_ID,
  [import_xy_chart_internals.LINE_VARIANT]: import_xy_chart_internals.LINE_ID,
  [import_xy_chart_internals.AREA_VARIANT]: import_xy_chart_internals.AREA_ID
};
function serializeBandArea(meta, point, xScale, yScale, xRelAcc, yRelAcc, yMin, yMax) {
  const coords = meta.orientation === "horizontal" ? [
    {
      x: (0, import_band_accessors.bandX0Accessor)(xScale, xRelAcc)(point.datapoint),
      y: (0, import_band_accessors.bandY0Accessor)(yScale, yRelAcc)(point.datapoint)
    },
    {
      x: (0, import_band_accessors.bandX1Accessor)(xScale, xRelAcc)(point.datapoint),
      y: (0, import_band_accessors.bandY0Accessor)(yScale, yRelAcc)(point.datapoint)
    }
  ] : [
    {
      x: (0, import_band_accessors.bandX0Accessor)(xScale, xRelAcc)(point.datapoint),
      y: (0, import_band_accessors.bandY0Accessor)(yScale, yRelAcc)(point.datapoint)
    },
    {
      x: (0, import_band_accessors.bandX0Accessor)(xScale, xRelAcc)(point.datapoint),
      y: (0, import_band_accessors.bandY1Accessor)(yScale, yRelAcc)(point.datapoint)
    }
  ];
  const visiblePoints = coords.filter(({ y }) => y >= yMin && y <= yMax).map(({ x, y }) => ({
    id: import_xy_chart_internals.AREA_ID,
    x: x + import_constants.POINT_OVERLAY_PADDING,
    y: y + import_constants.POINT_OVERLAY_PADDING,
    color: meta.color,
    shape: void 0
  }));
  return visiblePoints.length > 0 ? visiblePoints : null;
}
const serializeLinearSeriesValues = (linearMetadata, axisIdToScales, axisIdToScaleDomain, datapointMetadataMap) => (point) => {
  const meta = resolveMetadata(linearMetadata, point.series);
  if (!meta) {
    return null;
  }
  const xScale = (0, import_get_scale.getCurrentScale)(axisIdToScales, point.series.xAxisId);
  const yScale = (0, import_get_scale.getCurrentScale)(axisIdToScales, point.series.yAxisId);
  if (!xScale || !yScale) {
    return null;
  }
  const { xAxis, yAxis } = meta;
  const xRelAccessor = (0, import_value_representation.relAccessor)(xAxis, axisIdToScaleDomain);
  const yRelAccessor = (0, import_value_representation.relAccessor)(yAxis, axisIdToScaleDomain);
  const { xAccessor, yAccessor } = (0, import_get_linear_accessors.getLinearAccessors)(
    point.series.type,
    meta,
    datapointMetadataMap,
    xRelAccessor,
    yRelAccessor
  );
  const yBase = yAccessor(yScale)(point.datapoint);
  const [yMin, yMax] = getYRangeMinMax(yScale.range());
  if (yBase < yMin || yBase > yMax) {
    return null;
  }
  const areaMeta = point.series.type === import_xy_chart_internals.AREA_VARIANT ? meta : void 0;
  if (areaMeta?.variant === import_xy_chart_internals.BAND_VARIANT) {
    return serializeBandArea(
      areaMeta,
      point,
      xScale,
      yScale,
      xRelAccessor,
      yRelAccessor,
      yMin,
      yMax
    );
  }
  const xPos = xAccessor(xScale)(point.datapoint);
  return {
    id: idMap[point.series.type],
    x: xPos + import_constants.POINT_OVERLAY_PADDING,
    y: yBase + import_constants.POINT_OVERLAY_PADDING,
    color: meta.color,
    // shape is only defined for dot series
    shape: point.series.type === import_xy_chart_internals.DOT_VARIANT ? meta.shape : void 0
  };
};
