// @dynatrace/strato-components/charts/xy-chart/components/overlay/utils/serialize-rectangle-series-values.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var serialize_rectangle_series_values_exports = {};
__export(serialize_rectangle_series_values_exports, {
  serializeRectangleSeriesValues: () => serializeRectangleSeriesValues
});
module.exports = __toCommonJS(serialize_rectangle_series_values_exports);
var import_safe_get = require("../../../../core/utils/safe-get.js");
var import_xy_chart_internals = require("../../../types/xy-chart-internals.js");
var import_get_scale = require("../../../utils/get-scale.js");
var import_type_guards = require("../../../utils/type-guards.js");
var import_value_representation = require("../../../utils/value-representation.js");
var import_bar_accessors = require("../../displays/bar-accessors.js");
var import_BarDisplay = require("../../displays/BarDisplay/BarDisplay.js");
var import_rect_accessors = require("../../displays/rect-accessors.js");
const serializeRectangleSeriesValues = (axisIdToScales, rectMetadata, axisIdToScaleDomain, datapointMetadataMap, barMetadata) => {
  return (point) => {
    if ((0, import_type_guards.isRectDatapoint)(point.datapoint) && point.series.type === import_xy_chart_internals.RECT_VARIANT) {
      const currentXScale = (0, import_get_scale.getCurrentScale)(
        axisIdToScales,
        point.series.xAxisId
      );
      const currentYScale = (0, import_get_scale.getCurrentScale)(
        axisIdToScales,
        point.series.yAxisId
      );
      const { xAxis, yAxis } = (0, import_safe_get.safeGet)(rectMetadata, point.series);
      const xRelAccessor2 = (0, import_value_representation.relAccessor)(xAxis, axisIdToScaleDomain);
      const yRelAccessor2 = (0, import_value_representation.relAccessor)(yAxis, axisIdToScaleDomain);
      return {
        ...point,
        id: import_xy_chart_internals.RECT_ID,
        x: (0, import_rect_accessors.xAccessor)(currentXScale, xRelAccessor2)(point.datapoint),
        y: (0, import_rect_accessors.yAccessor)(currentYScale, yRelAccessor2)(point.datapoint),
        width: (0, import_rect_accessors.widthAccessor)(currentXScale, xRelAccessor2)(point.datapoint),
        height: (0, import_rect_accessors.heightAccessor)(
          currentYScale,
          yRelAccessor2
        )(point.datapoint)
      };
    }
    const xScale = (0, import_get_scale.getCurrentScale)(
      axisIdToScales,
      point.series.xAxisId,
      (0, import_BarDisplay.isXBaselineAccessor)(point.series.data[0]) ? point.series.type : void 0
    );
    const dimensionXScale = (0, import_get_scale.getDimensionScale)(
      axisIdToScales,
      point.series.xAxisId
    );
    const yScale = (0, import_get_scale.getCurrentScale)(
      axisIdToScales,
      point.series.yAxisId,
      (0, import_BarDisplay.isYBaselineAccessor)(point.series.data[0]) ? point.series.type : void 0
    );
    const dimensionYScale = (0, import_get_scale.getDimensionScale)(
      axisIdToScales,
      point.series.yAxisId
    );
    const barMeta = barMetadata ? (0, import_safe_get.safeGet)(barMetadata, point.series) : void 0;
    const xRelAccessor = barMeta ? (0, import_value_representation.relAccessor)(barMeta.xAxis, axisIdToScaleDomain) : void 0;
    const yRelAccessor = barMeta ? (0, import_value_representation.relAccessor)(barMeta.yAxis, axisIdToScaleDomain) : void 0;
    const x = (0, import_bar_accessors.xAccessor)(
      xScale,
      datapointMetadataMap,
      dimensionXScale,
      point.series.id,
      xRelAccessor
    )(point.datapoint);
    const y = (0, import_bar_accessors.yAccessor)(
      yScale,
      datapointMetadataMap,
      dimensionYScale,
      point.series.id,
      yRelAccessor
    )(point.datapoint);
    const width = (0, import_bar_accessors.widthAccessor)(
      xScale,
      datapointMetadataMap,
      dimensionXScale,
      xRelAccessor
    )(point.datapoint);
    const height = (0, import_bar_accessors.heightAccessor)(
      yScale,
      datapointMetadataMap,
      dimensionYScale,
      yRelAccessor
    )(point.datapoint);
    return {
      ...point,
      id: import_xy_chart_internals.BAR_ID,
      x,
      y,
      width,
      height
    };
  };
};
