// @dynatrace/strato-components/charts/core/components/overlay-tooltip/components/DynamicTooltip.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DynamicTooltip_exports = {};
__export(DynamicTooltip_exports, {
  DynamicTooltip: () => DynamicTooltip
});
module.exports = __toCommonJS(DynamicTooltip_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_TooltipPortal = require("./TooltipPortal.js");
var import_useThrottledCallbackOnResize = require("../../../hooks/useThrottledCallbackOnResize.js");
var import_useThrottledCallbackOnScroll = require("../../../hooks/useThrottledCallbackOnScroll.js");
var import_throttleRaf = require("../../../utils/throttleRaf.js");
var import_OverlayContext = require("../contexts/OverlayContext.js");
var import_TooltipContext = require("../contexts/TooltipContext.js");
var import_useTooltipReducer = require("../useTooltipReducer.js");
const DynamicTooltip = (props) => {
  const { absoluteX, absoluteY, pinned } = props;
  const overlay = (0, import_OverlayContext.useOverlay)();
  const dispatch = (0, import_useTooltipReducer.useTooltipReducer)();
  const tooltip = (0, import_TooltipContext.useTooltip)();
  const onEscapeKeyPressedHandler = (0, import_react.useCallback)((event) => {
    if (event.key === "Escape") {
      dispatch({ type: "RESET_TOOLTIP" })();
    }
  }, []);
  (0, import_react.useEffect)(() => {
    window.addEventListener("keyup", onEscapeKeyPressedHandler);
    return () => {
      window.removeEventListener("keyup", onEscapeKeyPressedHandler);
    };
  }, []);
  (0, import_useThrottledCallbackOnScroll.useThrottledCallbackOnScroll)(false, () => {
    if (!overlay.overlayRef.current) {
      return;
    }
    if (tooltip.getState().pinned) {
      dispatch({
        type: "UPDATE_POSITION",
        payload: { type: "non-resize", scroll: true }
      })();
      return;
    }
    dispatch({ type: "SET_OUT_BOUNDS" })();
  });
  (0, import_useThrottledCallbackOnResize.useThrottledCallbackOnResize)(
    overlay.overlayRef,
    dispatch({ type: "UPDATE_POSITION", payload: { type: "resize" } }),
    !pinned
  );
  const pointerMoveCallback = (0, import_react.useMemo)(
    () => (0, import_throttleRaf.throttleRaf)((event) => {
      if (event.pressure && tooltip.getState().pinned) {
        if (!overlay.overlayRef.current || overlay.overlayRef.current.contains(event.target)) {
          return;
        }
        dispatch({
          type: "UPDATE_POSITION",
          payload: { type: "non-resize" }
        })();
        return;
      }
    }),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    []
  );
  const pointerUpCallback = (0, import_react.useMemo)(
    () => (0, import_throttleRaf.throttleRaf)((event) => {
      if (tooltip.getState().pinned) {
        if (!overlay.overlayRef.current || overlay.overlayRef.current.contains(event.target)) {
          return;
        }
        dispatch({
          type: "UPDATE_POSITION",
          payload: { type: "non-resize" }
        })();
        return;
      }
    }),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    []
  );
  (0, import_react.useEffect)(() => {
    if (!pinned || !overlay.overlayRef.current) {
      return;
    }
    window.addEventListener("pointermove", pointerMoveCallback);
    window.addEventListener("pointerup", pointerUpCallback);
    return () => {
      window.removeEventListener("pointermove", pointerMoveCallback);
      window.removeEventListener("pointerup", pointerUpCallback);
    };
  }, [pinned]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_TooltipPortal.TooltipPortal,
    {
      isInteractive: pinned,
      absoluteX,
      absoluteY,
      onMouseLeave: dispatch({ type: "TOOLTIP_MOUSE_LEAVE" }),
      onScrollCapture: () => dispatch({ type: "TOOLTIP_MOUSE_LEAVE" }),
      children: props.children
    }
  );
};
