// @dynatrace/strato-components/charts/core/components/overlay-tooltip/components/TooltipPortal.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TooltipPortal_exports = {};
__export(TooltipPortal_exports, {
  TooltipPortal: () => TooltipPortal
});
module.exports = __toCommonJS(TooltipPortal_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_dom = require("react-dom");
var import_OverlayStackingProvider = require("../../../../../core/providers/OverlayStackingProvider.js");
var import_useDynamicTooltip = require("../hooks/useDynamicTooltip.js");
var import_tooltip_container_dimensions_Provider = require("../providers/tooltip-container-dimensions.Provider.js");
var import_tooltip_css = require("../tooltip.sty.js");
const TooltipPortal = (props) => {
  const { absoluteX, absoluteY, isInteractive, onMouseLeave, onScrollCapture } = props;
  const [tooltipWrapper, setTooltipWrapper] = (0, import_react.useState)(
    null
  );
  const dynamicTooltip = (0, import_useDynamicTooltip.useDynamicTooltip)({
    template: tooltipWrapper,
    x: absoluteX,
    y: absoluteY
  });
  return (0, import_react_dom.createPortal)(
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "div",
      {
        role: "tooltip",
        "aria-label": "OverlayTooltip",
        ref: setTooltipWrapper,
        className: (0, import_tooltip_css.wrapper)({ isInteractive }),
        onMouseLeave,
        onScrollCapture,
        ...dynamicTooltip,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayStackingProvider.OverlayStackingProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_tooltip_container_dimensions_Provider.TooltipContainerDimensionProvider, { container: tooltipWrapper, children: props.children }) })
      }
    ),
    document.body
  );
};
