// @dynatrace/strato-components/charts/xy-chart/components/displays/BarDisplay/BarDisplay.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var BarDisplay_exports = {};
__export(BarDisplay_exports, {
  BarDisplay: () => BarDisplay,
  isXBaselineAccessor: () => isXBaselineAccessor,
  isYBaselineAccessor: () => isYBaselineAccessor
});
module.exports = __toCommonJS(BarDisplay_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_BarCanvas = require("../../../../core/components/canvas/components/BarCanvas.js");
var import_useGridPlotSize = require("../../../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_constants = require("../../../../core/components/data-point/constants.js");
var import_constants2 = require("../../../../core/constants.js");
var import_safe_get = require("../../../../core/utils/safe-get.js");
var import_useBarData = require("../../../hooks/useBarData.js");
var import_useXYChartData = require("../../../hooks/useXYChartData.js");
var import_get_scale = require("../../../utils/get-scale.js");
var import_value_representation = require("../../../utils/value-representation.js");
var import_bar_accessors = require("../bar-accessors.js");
function isXBaselineAccessor(datapoint) {
  return !!datapoint && datapoint.x0 !== void 0;
}
function isYBaselineAccessor(datapoint) {
  return !!datapoint && datapoint.y0 !== void 0;
}
function BarDisplay({
  axisIdToScales,
  series,
  metadata
}) {
  const { height, width } = (0, import_useGridPlotSize.useGridPlotSize)();
  const xScale = (0, import_get_scale.getCurrentScale)(
    axisIdToScales,
    series.xAxisId,
    isXBaselineAccessor(series.data[0]) ? series.type : void 0
  );
  const dimensionXScale = (0, import_get_scale.getDimensionScale)(axisIdToScales, series.xAxisId);
  const yScale = (0, import_get_scale.getCurrentScale)(
    axisIdToScales,
    series.yAxisId,
    isYBaselineAccessor(series.data[0]) ? series.type : void 0
  );
  const dimensionYScale = (0, import_get_scale.getDimensionScale)(axisIdToScales, series.yAxisId);
  const { color } = metadata;
  const { datapointMetadataMap, axisIdToScaleDomain } = (0, import_useXYChartData.useXYChartData)();
  const { seriesToMetadata } = (0, import_useBarData.useBarData)();
  const { xAxis, yAxis } = (0, import_safe_get.safeGet)(seriesToMetadata, series);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_BarCanvas.BarCanvas,
    {
      data: series.data,
      canvasHeight: height,
      canvasWidth: width,
      opacity: () => import_constants2.DEFAULT_FILL_OPACITY,
      color: () => color.Default,
      x: (0, import_bar_accessors.xAccessor)(
        xScale,
        datapointMetadataMap,
        dimensionXScale,
        series.id,
        (0, import_value_representation.relAccessor)(xAxis, axisIdToScaleDomain)
      ),
      y: (0, import_bar_accessors.yAccessor)(
        yScale,
        datapointMetadataMap,
        dimensionYScale,
        series.id,
        (0, import_value_representation.relAccessor)(yAxis, axisIdToScaleDomain)
      ),
      width: (0, import_bar_accessors.widthAccessor)(
        xScale,
        datapointMetadataMap,
        dimensionXScale,
        (0, import_value_representation.relAccessor)(xAxis, axisIdToScaleDomain)
      ),
      height: (0, import_bar_accessors.heightAccessor)(
        yScale,
        datapointMetadataMap,
        dimensionYScale,
        (0, import_value_representation.relAccessor)(yAxis, axisIdToScaleDomain)
      ),
      enableDotOverflow: true,
      overlayPadding: import_constants.POINT_OVERLAY_PADDING
    },
    series.id
  );
}
BarDisplay["displayName"] = "BarDisplay";
