// @dynatrace/strato-components/charts/xy-chart/components/displays/BarDisplay/BarLabelLayout.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var BarLabelLayout_exports = {};
__export(BarLabelLayout_exports, {
  BarLabelLayout: () => BarLabelLayout
});
module.exports = __toCommonJS(BarLabelLayout_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_BarLabelOverlay = require("./BarLabelOverlay.js");
var import_ChartGrid_css = require("../../../../core/components/chart-grid/ChartGrid.sty.js");
var import_xy_chart_z_index_areas = require("../../../constants/xy-chart-z-index-areas.js");
var import_useBarData = require("../../../hooks/useBarData.js");
var import_useXYChartScales = require("../../../hooks/useXYChartScales.js");
function BarLabelLayout() {
  const axisIdToScales = (0, import_useXYChartScales.useXYChartScales)();
  const { filteredSeriesWithDatapoints, seriesToMetadata } = (0, import_useBarData.useBarData)();
  const labelSeries = filteredSeriesWithDatapoints.filter(
    (s) => s.hasLabels === true
  );
  if (labelSeries.length === 0) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      className: import_ChartGrid_css.plotLayout,
      style: {
        zIndex: import_xy_chart_z_index_areas.zIndicesXYChartAreas.plot + 1,
        pointerEvents: "none"
      },
      children: labelSeries.map((series) => {
        const metadata = seriesToMetadata.get(series);
        if (!metadata) {
          return null;
        }
        return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_BarLabelOverlay.BarLabelOverlay,
          {
            axisIdToScales,
            series,
            metadata
          },
          series.id
        );
      })
    }
  );
}
BarLabelLayout["displayName"] = "BarLabelLayout";
