// @dynatrace/strato-components/charts/xy-chart/components/displays/BarDisplay/HorizontalBarLabel.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var HorizontalBarLabel_exports = {};
__export(HorizontalBarLabel_exports, {
  HorizontalBarLabel: () => HorizontalBarLabel
});
module.exports = __toCommonJS(HorizontalBarLabel_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_BarLabelOverlay_css = require("./BarLabelOverlay.sty.js");
var import_label_utils = require("./label-utils.js");
var import_chars_font_width_map = require("../../../../core/components/text-ellipsis/chars-font-width-map.js");
var import_truncate_string = require("../../../../core/components/text-ellipsis/truncate-string.js");
function HorizontalBarLabel({
  x,
  y,
  barWidth,
  barHeight,
  canvasWidth,
  isReversed,
  text,
  insideColor,
  outsideColor,
  truncationMode
}) {
  if (barHeight < import_label_utils.BAR_LABEL_MIN_SIZE) {
    return null;
  }
  const centerY = y + barHeight / 2;
  const availLength = isReversed ? x + barWidth - 2 * import_label_utils.BAR_LABEL_PADDING : canvasWidth - x - 2 * import_label_utils.BAR_LABEL_PADDING;
  if (availLength < import_label_utils.BAR_LABEL_MIN_SIZE) {
    return null;
  }
  const truncated = (0, import_truncate_string.truncateString)(
    text,
    availLength,
    import_chars_font_width_map.charsFontWidth12pixelsSize400Map,
    truncationMode
  );
  const anchor = isReversed ? { right: import_label_utils.BAR_LABEL_PADDING, textAlign: "right" } : { left: x + import_label_utils.BAR_LABEL_PADDING };
  const outerStyle = {
    ...anchor,
    top: centerY,
    transform: "translateY(-50%)",
    width: availLength,
    color: outsideColor
  };
  const wrapperStyle = {
    left: x,
    top: y,
    width: barWidth,
    height: barHeight
  };
  const innerStyle = {
    ...isReversed ? anchor : { left: import_label_utils.BAR_LABEL_PADDING },
    top: barHeight / 2,
    transform: "translateY(-50%)",
    width: availLength,
    color: insideColor
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_react.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { className: import_BarLabelOverlay_css.labelSpan, style: outerStyle, children: truncated }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: import_BarLabelOverlay_css.labelClipWrapper, style: wrapperStyle, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { className: import_BarLabelOverlay_css.labelSpan, style: innerStyle, children: truncated }) })
  ] });
}
HorizontalBarLabel["displayName"] = "HorizontalBarLabel";
