// @dynatrace/strato-components/forms/password-input/PasswordInput.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var PasswordInput_exports = {};
__export(PasswordInput_exports, {
  PasswordInput: () => PasswordInput
});
module.exports = __toCommonJS(PasswordInput_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_PasswordInputTooltip = require("./PasswordInputTooltip.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useControllableState = require("../../core/hooks/useControllableState.js");
var import_extract_behavior_tracking_props = require("../../core/utils/extract-behavior-tracking-props.js");
var import_is_slot = require("../../core/utils/is-slot.js");
var import_merge_props = require("../../core/utils/merge-props.js");
var import_BaseInput = require("../_base-input/BaseInput.js");
var import_baseInputMessages = require("../_base-input/baseInputMessages.js");
const COMPONENT_NAME = "PasswordInput";
function getTooltipSlot(children) {
  const slots = {
    tooltip: null
  };
  for (const child of import_react.Children.toArray(children)) {
    if ((0, import_is_slot.isSlot)(child, import_PasswordInputTooltip.PasswordInputTooltip)) {
      slots.tooltip = child;
    } else if (child !== null) {
      console.warn("The used child is not supported by the PasswordInput.");
    }
  }
  return slots;
}
const PasswordInputRoot = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const intl = (0, import_react_intl.useIntl)();
  const defaultPlaceholder = intl.formatMessage(
    (0, import_baseInputMessages.getPlaceholderMessage)("password")
  );
  const {
    onChange,
    name,
    readOnly = false,
    disabled,
    controlState,
    children,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "aria-disabled": ariaDisabledProp,
    className: consumerClassName,
    style: consumerStyle,
    defaultValue,
    value,
    placeholder = defaultPlaceholder,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const [hidden, setHidden] = (0, import_react.useState)(true);
  const [internalValue, setInternalValue] = (0, import_useControllableState.useControllableState)({
    defaultValue,
    value
  });
  const onInternalChange = (event) => {
    onChange?.(event.target.value, event);
    setInternalValue(event.target.value);
  };
  const { tooltip } = getTooltipSlot(children);
  const { showText, hideText, ...remainingTooltipProps } = tooltip?.props ?? {};
  const tooltipBehavioralTrackingProps = (0, import_extract_behavior_tracking_props.extractBehaviorTrackingProps)(
    remainingTooltipProps
  );
  const showTooltipDefaultText = intl.formatMessage({
    defaultMessage: "Show password",
    id: "tgL6QEjmq+ArASFk",
    description: "Text in a tooltip shown when hovering the password input icon to show the password"
  });
  const hideTooltipDefaultText = intl.formatMessage({
    defaultMessage: "Hide password",
    id: "vMt3ZRHwBods5xzM",
    description: "Text in a tooltip shown when hovering the password input icon to hide the password"
  });
  const dataDtProperties = `value:${hidden ? "show" : "hide"}`;
  const showTooltipText = showText ?? showTooltipDefaultText;
  const hideTooltipText = hideText ?? hideTooltipDefaultText;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_BaseInput.BaseInput,
    {
      ...(0, import_merge_props.mergeProps)(restProps, behavioralTrackingProps),
      value: internalValue,
      role: "textbox",
      controlState,
      name,
      type: hidden ? "password" : "text",
      onChange: onInternalChange,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ref: forwardedRef,
      className: consumerClassName,
      style: consumerStyle,
      readOnly,
      disabled,
      "aria-disabled": ariaDisabledProp,
      "data-dt-component": COMPONENT_NAME,
      placeholder,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_BaseInput.BaseInput.Suffix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_BaseInput.BaseInput.Button,
        {
          onClick: () => setHidden(!hidden),
          tooltipText: hidden ? showTooltipText : hideTooltipText,
          tooltipBehavioralTrackingProps,
          "aria-label": hidden ? showTooltipText : hideTooltipText,
          id: remainingTooltipProps.id,
          "aria-disabled": ariaDisabledProp,
          "data-dt-name": "ShowHidePassword",
          "data-dt-properties": dataDtProperties,
          disabled: readOnly ? false : disabled,
          children: hidden ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ViewIcon, {}) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ViewOffIcon, {})
        }
      ) })
    }
  );
});
const PasswordInput = Object.assign(PasswordInputRoot, {
  Tooltip: import_PasswordInputTooltip.PasswordInputTooltip
});
PasswordInput.displayName = COMPONENT_NAME;
