// @dynatrace/strato-components/charts/core/components/sparkline-display/utils/accessors.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var accessors_exports = {};
__export(accessors_exports, {
  barHeightAccessor: () => barHeightAccessor,
  barWidthAccessor: () => barWidthAccessor,
  barXAxisAccessor: () => barXAxisAccessor,
  barYAxisAccessor: () => barYAxisAccessor,
  definedAccessor: () => definedAccessor,
  isIsolatedAccessor: () => isIsolatedAccessor,
  xAxisAccessor: () => xAxisAccessor,
  y0Accessor: () => y0Accessor,
  y1Accessor: () => y1Accessor,
  yAccessor: () => yAccessor
});
module.exports = __toCommonJS(accessors_exports);
var import_is_isolated_point = require("../../../utils/is-isolated-point.js");
var import_sanitize_zero_values = require("../../../utils/sanitize-zero-values.js");
var import_constants = require("../constants.js");
const xAxisAccessor = (xScale) => (dataPoint) => xScale(dataPoint.end ?? dataPoint.start) ?? 0;
const yAccessor = (yScale) => (dataPoint) => yScale(dataPoint.value) ?? 0;
const y0Accessor = (yScale) => () => (0, import_sanitize_zero_values.sanitizeZeroValues)(yScale, 0) ?? 0;
const y1Accessor = (yScale) => (dataPoint) => yScale(dataPoint.value) ?? 0;
const definedAccessor = (datapoint) => {
  return !datapoint.isGap;
};
const isIsolatedAccessor = (dataPoints, index) => (0, import_is_isolated_point.isIsolatedPoint)(dataPoints, index, (point) => definedAccessor(point));
const barXAxisAccessor = (xScale) => (dataPoint) => {
  const { start, end } = dataPoint;
  const isTimeframe = end && end !== start;
  const deviation = isTimeframe ? 0 : import_constants.TIMESTAMP_BAR_X_DEVIATION;
  return (xScale(start) ?? 0) - deviation;
};
const barYAxisAccessor = (yScale) => (dataPoint) => {
  const maxValue = yScale.invert(0);
  const maxValueCoordinate = (0, import_sanitize_zero_values.sanitizeZeroValues)(yScale, maxValue);
  const cumulativeCoordinate = (0, import_sanitize_zero_values.sanitizeZeroValues)(yScale, dataPoint.value);
  return maxValueCoordinate + cumulativeCoordinate;
};
const barWidthAccessor = (datapoint, xScale) => {
  const { start, end } = datapoint;
  const isTimeframe = end && end !== start;
  return isTimeframe ? (xScale(end) ?? 0) - (xScale(start) ?? 0) : import_constants.TIMESTAMP_BAR_WIDTH;
};
const barHeightAccessor = (datapoint, yScale) => {
  const valueCoordinate = (0, import_sanitize_zero_values.sanitizeZeroValues)(yScale, datapoint.value);
  const zeroCoordinate = (0, import_sanitize_zero_values.sanitizeZeroValues)(yScale, 0);
  return zeroCoordinate - valueCoordinate;
};
