// @dynatrace/strato-components/charts/pie/components/PieChartOverlays/PieChartOverlays.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var PieChartOverlays_exports = {};
__export(PieChartOverlays_exports, {
  PieChartOverlays: () => PieChartOverlays
});
module.exports = __toCommonJS(PieChartOverlays_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_PieChartOverlays_css = require("./PieChartOverlays.sty.js");
var import_Flex = require("../../../../layouts/flex/Flex.js");
var import_useChartLayoutLoading = require("../../../core/components/chart-layout/hooks/useChartLayoutLoading.js");
var import_OverlayContext = require("../../../core/components/overlay-tooltip/contexts/OverlayContext.js");
var import_is_toolbar_enabled = require("../../../core/utils/is-toolbar-enabled.js");
var import_ToolbarContext = require("../../context/ToolbarContext.js");
var import_useOverlayHandlers = require("../../hooks/useOverlayHandlers.js");
var import_PieChartToolbar = require("../PieChartToolbar/PieChartToolbar.js");
var import_PieChartTooltip = require("../PieChartTooltip/PieChartTooltip.js");
const PieChartOverlays = ({ children }) => {
  const overlay = (0, import_OverlayContext.useOverlay)();
  const loading = (0, import_useChartLayoutLoading.useChartLayoutLoading)();
  const { onMouseEnterHandler, onMouseLeaveHandler } = (0, import_useOverlayHandlers.useOverlayHandlers)();
  const toolbarConfig = (0, import_ToolbarContext.useToolbarConfig)();
  const toolbarEnabled = (0, import_is_toolbar_enabled.isToolbarEnabled)(toolbarConfig);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_Flex.Flex,
    {
      ref: overlay.overlayRef,
      "data-testid": "overlays-wrapper",
      justifyContent: "center",
      alignItems: "center",
      className: import_PieChartOverlays_css.container,
      onMouseEnter: loading ? void 0 : onMouseEnterHandler,
      onMouseLeave: loading ? void 0 : onMouseLeaveHandler,
      children: [
        !loading && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_PieChartTooltip.PieChartTooltip, {}),
        !loading && toolbarEnabled && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_PieChartToolbar.PieChartToolbar, {}),
        children
      ]
    }
  );
};
