// @dynatrace/strato-components/charts/histogram/components/overlays/constants.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var constants_exports = {};
__export(constants_exports, {
  ADJUSTED_LAYOUT_OFFSET: () => ADJUSTED_LAYOUT_OFFSET,
  PREPOSITION: () => PREPOSITION,
  leftMouseClick: () => leftMouseClick,
  middleMouseClick: () => middleMouseClick
});
module.exports = __toCommonJS(constants_exports);
var import_react_intl = require("react-intl");
const leftMouseClick = 1;
const middleMouseClick = 4;
const ADJUSTED_LAYOUT_OFFSET = { x: -10, y: 30 };
const PREPOSITION = (0, import_react_intl.defineMessages)({
  to: {
    id: "Sovq/8OI4lXhwuJJ",
    defaultMessage: "to",
    description: "from / to"
  }
});
