// @dynatrace/strato-components/charts/histogram/components/overlays/HistogramOverlays.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var HistogramOverlays_exports = {};
__export(HistogramOverlays_exports, {
  HistogramOverlays: () => HistogramOverlays
});
module.exports = __toCommonJS(HistogramOverlays_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_HistogramInspect = require("./HistogramExplore/HistogramInspect.js");
var import_useApplyPanDelta = require("./hooks/useApplyPanDelta.js");
var import_useApplyZoomDelta = require("./hooks/useApplyZoomDelta.js");
var import_useGridPlotSize = require("../../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_InteractionsOverlay = require("../../../core/components/interactions-overlay/InteractionsOverlay.js");
var import_types = require("../../../core/components/interactions-overlay/types.js");
var import_is_own_event = require("../../../core/components/interactions-overlay/utils/is-own-event.js");
var import_explore_reducers = require("../../../core/components/interactions-state-helpers/reducers/explore-reducers.js");
var import_mode_reducers = require("../../../core/components/interactions-state-helpers/reducers/mode-reducers.js");
var import_useToolbarHandlerRef = require("../../../core/components/toolbar/hooks/useToolbarHandlerRef.js");
var import_get_percentage_from_pixel = require("../../../core/utils/get-percentage-from-pixel.js");
var import_HistogramChartOverlayRefContext = require("../../contexts/HistogramChartOverlayRefContext.js");
var import_HistogramSelectionAreaRef_context = require("../../contexts/HistogramSelectionAreaRef.context.js");
var import_selectors = require("../../hooks/store/selectors.js");
var import_state_setters = require("../../hooks/store/state-setters.js");
var import_useHistogramInteractions = require("../../hooks/useHistogramInteractions.js");
var import_useHistogramToolbar = require("../../hooks/useHistogramToolbar.js");
var import_useResetInteractions = require("../../hooks/useResetInteractions.js");
var import_useResetOnResize = require("../../hooks/useResetOnResize.js");
var import_useZoomToSelection = require("../../hooks/useZoomToSelection.js");
var import_HistogramToolbar = require("../toolbar/HistogramToolbar.js");
var import_HistogramCrosshair = require("./HistogramExplore/HistogramCrosshair.js");
var import_HistogramSelectionArea = require("./HistogramSelectionArea/HistogramSelectionArea.js");
var import_useSetClosestBins = require("../../hooks/useSetClosestBins.js");
const HistogramOverlays = () => {
  const handlerRef = (0, import_useToolbarHandlerRef.useToolbarHandlerRef)();
  const { zoomX, pan, zoom } = (0, import_useHistogramInteractions.useHistogramInteractions)();
  const selectionRef = (0, import_react.useRef)(null);
  const { width, height } = (0, import_useGridPlotSize.useGridPlotSize)();
  const toolbar = (0, import_useHistogramToolbar.useHistogramToolbar)();
  const setState = (0, import_state_setters.useSetState)();
  const mode = (0, import_selectors.useInteractionModeState)();
  const exploreState = (0, import_selectors.useExploreState)();
  const selectState = (0, import_selectors.useSelectState)();
  const zoomState = (0, import_selectors.useZoomState)();
  const overlayRef = (0, import_react.useRef)(null);
  const applyPanDelta = (0, import_useApplyPanDelta.useApplyPanDelta)();
  const applyZoomDelta = (0, import_useApplyZoomDelta.useApplyZoomDelta)();
  const zoomToSelection = (0, import_useZoomToSelection.useZoomToSelection)();
  const resetInteractions = (0, import_useResetInteractions.useResetInteractions)();
  const setClosestBins = (0, import_useSetClosestBins.useSetClosestBins)();
  (0, import_useResetOnResize.useResetOnResize)();
  const handleModeChange = (event) => {
    const canPan = zoomState.zoomOutEnabled;
    const shouldSwitchToPanMode = event.mode === "pan" && pan?.enabled && canPan;
    const shouldSwitchToZoomMode = event.mode === "zoom" && (zoomX?.enabled || zoom?.enabled);
    if (shouldSwitchToPanMode || shouldSwitchToZoomMode) {
      setState((0, import_mode_reducers.reduceInteractionMode)(event.mode));
      setState(import_explore_reducers.reduceExploreReset);
    } else if (event.mode === "explore") {
      setState((0, import_mode_reducers.reduceInteractionMode)(import_types.INTERACTION_MODE.Explore));
    }
  };
  const handleFocus = (event) => {
    if (!event || event.cancelable) {
      return;
    }
    if ((0, import_is_own_event.isOwnEvent)(event)) {
      handlerRef.current.handleFocus();
    }
  };
  const handleHover = (event) => {
    if (mode === "explore") {
      setClosestBins(event.position);
      setState((0, import_explore_reducers.reduceExplorePosition)(event.position));
    }
  };
  const handleLeave = (_) => {
    if (mode === "explore") {
      setState(import_explore_reducers.reduceExploreEnd);
    }
  };
  const handlePanChange = (event) => {
    if (pan?.enabled) {
      applyPanDelta(event.delta);
    }
  };
  const handleTogglePin = (event) => {
    if (event.pin) {
      const highlightedSeriesId = setClosestBins(event.position);
      setState((0, import_explore_reducers.reduceExplorePin)(event.position, highlightedSeriesId));
    } else {
      setState((0, import_explore_reducers.reduceExploreUnPin)(event.position));
    }
  };
  const handleReset = (event) => {
    resetInteractions(event.mousePosition);
  };
  const handleSelectStart = (event) => {
    if (!zoomX?.enabled) {
      handleTogglePin({
        triggeredBy: event.triggeredBy,
        position: event.position,
        pin: !exploreState.pinned
      });
    }
  };
  const handleSelectChange = (event) => {
    if (zoomX?.enabled || mode === "explore") {
      const variant = event.mode === "explore" ? "select" : "zoom";
      setState(
        (0, import_explore_reducers.reduceExploreSelect)({
          start: event.start,
          end: event.end,
          finished: false,
          accessible: false,
          variant,
          selectionPercentage: (0, import_get_percentage_from_pixel.getPercentageFromPixel)(
            [event.start, event.end],
            width
          )
        })
      );
    }
  };
  const handleSelectEnd = (event) => {
    if ((0, import_types.isSelectCancelled)(event)) {
      setState(import_explore_reducers.reduceExploreCancelSelect);
    } else if (mode === "explore") {
      setState(
        (0, import_explore_reducers.reduceExploreSelect)({
          start: event.start,
          end: event.end,
          finished: true,
          accessible: false,
          selectionPercentage: (0, import_get_percentage_from_pixel.getPercentageFromPixel)(
            [event.start, event.end],
            width
          )
        })
      );
      selectionRef.current?.focus();
    } else if (zoomX?.enabled) {
      zoomToSelection();
    }
  };
  const handleZoomChange = (event) => {
    applyZoomDelta(event.deltaStart, event.deltaEnd, event.behavior);
    setState(import_explore_reducers.reduceExploreCancelAll);
  };
  const handleToolbarMouseEnter = () => {
    setState(import_explore_reducers.reduceExploreCancelTooltip);
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_InteractionsOverlay.InteractionsOverlay,
    {
      ref: overlayRef,
      mode,
      pinned: exploreState.pinned,
      selectionArea: selectState ? [selectState?.start, selectState?.end] : void 0,
      width,
      height,
      isZoomDisabled: !zoom?.enabled,
      isZoomXDisabled: !zoomX?.enabled,
      isPanDisabled: !(pan?.enabled && zoomState.zoomOutEnabled),
      onModeChange: handleModeChange,
      onFocus: handleFocus,
      onHover: handleHover,
      onLeave: handleLeave,
      onPanChange: handlePanChange,
      onTogglePin: handleTogglePin,
      onReset: handleReset,
      onSelectStart: handleSelectStart,
      onSelectChange: handleSelectChange,
      onSelectEnd: handleSelectEnd,
      onZoomChange: handleZoomChange,
      "data-testid": "histogram-chart-overlay",
      style: { userSelect: selectState?.finished === false ? "none" : "auto" },
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_HistogramChartOverlayRefContext.HistogramChartOverlayRefContext.Provider, { value: overlayRef, children: [
        exploreState.visible ? /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_HistogramCrosshair.HistogramCrosshair, {}),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_HistogramInspect.HistogramInspect, {})
        ] }) : null,
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_HistogramSelectionAreaRef_context.HistogramSelectionAreaRef.Provider, { value: selectionRef, children: [
          toolbar.enabled ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_HistogramToolbar.HistogramToolbar,
            {
              onMouseEnter: handleToolbarMouseEnter,
              handlerRef
            }
          ) : null,
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_HistogramSelectionArea.HistogramSelectionArea, {})
        ] })
      ] })
    }
  );
};
