// @dynatrace/strato-components/charts/histogram/components/overlays/HistogramDatapoints.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var HistogramDatapoints_exports = {};
__export(HistogramDatapoints_exports, {
  HistogramDatapoints: () => HistogramDatapoints
});
module.exports = __toCommonJS(HistogramDatapoints_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_HistogramOverlays_css = require("./HistogramOverlays.sty.js");
var import_Canvas = require("../../../core/components/canvas/Canvas.js");
var import_CanvasContext = require("../../../core/components/canvas/CanvasContext.js");
var import_useGridPlotSize = require("../../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_constants = require("../../../core/components/data-point/constants.js");
var import_DataPointCanvas = require("../../../core/components/data-point/DataPointCanvas.js");
const HistogramDatapoints = (0, import_react.memo)(
  (props) => {
    const { closestPoint, groupedPoints } = props;
    const { width, height } = (0, import_useGridPlotSize.useGridPlotSize)();
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: import_HistogramOverlays_css.datapointLayout, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_CanvasContext.CanvasContext.Provider,
      {
        value: {
          shapeRefs: /* @__PURE__ */ new Map()
        },
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_Canvas.Canvas,
          {
            width: width + import_constants.OVERLAY_PADDING * 2,
            height: height + import_constants.OVERLAY_PADDING * 2,
            children: groupedPoints.length > 0 && groupedPoints.map((pointGroup, index) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_DataPointCanvas.DataPointCanvas,
              {
                points: pointGroup,
                closestPoint
              },
              `pointGroup-${index}`
            ))
          }
        )
      }
    ) });
  }
);
HistogramDatapoints["displayName"] = "HistogramDatapoints";
