// @dynatrace/strato-components/esm/filters/filter-field/utils/conversion/lezer-tree-conversion/value/free-text.js
import { postSpotErrorMark } from "../../../../decorations/decoration-marks.js";
import { getTextFromRange } from "../../../get-text.js";
import { getColoredSearchOperator } from "../../../validation-utils/validation-helper-functions.js";
import { removeEscapeCharacters } from "../../escape-characters.js";
function convertFreeText(node, doc, parserConfig) {
  const { from, to } = node;
  const text = getTextFromRange(node, doc);
  const value = removeEscapeCharacters(text);
  const errors = [];
  const convertedNode = {
    type: "Error",
    range: {
      from,
      to
    },
    textValue: text,
    value
  };
  if (parserConfig?.searchConversion) {
    errors.push({
      range: { from: to - 1, to },
      message: {
        type: "incompleteStatementWithEnabledSearchError",
        details: getColoredSearchOperator()
      },
      decoration: postSpotErrorMark
    });
  } else {
    errors.push({
      range: { from: to - 1, to },
      message: "incompleteStatementError",
      decoration: postSpotErrorMark
    });
  }
  return {
    convertedNode,
    errors
  };
}
export {
  convertFreeText
};
