// @dynatrace/strato-components/esm/filters/filter-field/utils/conversion/lezer-tree-conversion/value/identifier.js
import { getTextFromRange } from "../../../get-text.js";
import {
  removeEscapeCharacters,
  isValidEscapeSequence
} from "../../escape-characters.js";
function convertIdentifier(node, doc, context) {
  const { from, to } = node;
  const text = getTextFromRange(node, doc);
  const value = removeEscapeCharacters(
    context === "String" ? text.slice(1, -1) : text
  );
  const convertedNode = {
    type: "String",
    range: {
      from,
      to
    },
    textValue: text,
    value
  };
  let endsWithOperator;
  let startsWithOperator;
  if (node.name === "String") {
    endsWithOperator = text.slice(1, 2) === "*" ? { from: from + 1, to: from + 2 } : void 0;
    const withoutQuotes = text.slice(1, text.length - 1);
    startsWithOperator = withoutQuotes.match(/(?<!\\)(?:\\\\)*[*]$/) ? { from: to - 2, to: to - 1 } : void 0;
    convertedNode.value = value.substring(
      endsWithOperator ? 1 : 0,
      value.length - (startsWithOperator ? 1 : 0)
    );
  }
  return {
    convertedNode,
    errors: [],
    isValid: isValidEscapeSequence(text),
    positionOperators: {
      endsWithOperator,
      startsWithOperator
    }
  };
}
export {
  convertIdentifier
};
