// @dynatrace/strato-components/esm/filters/filter-field/utils/conversion/lezer-tree-conversion/value/simple-value.js
import { isDev } from "../../../../../../core/utils/is-node-environment.js";
import { getTextFromRange } from "../../../get-text.js";
import {
  isValidIPAddress,
  isValidSmartscapeId,
  isValidTimestamp,
  isValidUID
} from "../../../validation-utils/type-validators.js";
import { convertNode } from "../generic-node.js";
import { convertToErrorNode, updateNodeValues } from "../node-mutations.js";
import { convertBoolean } from "./boolean.js";
import { convertDuration } from "./duration.js";
import { convertIdentifier } from "./identifier.js";
import { convertNumber } from "./number.js";
import { isValidCustomType } from "../../../validation-utils/validation-helper-functions.js";
function convertSimpleValue(node, doc, numberParser, customTypes, parserConfig) {
  const { from, to } = node;
  const cursor = node.node.cursor();
  let convertedNode = {
    type: "String",
    range: {
      from,
      to
    },
    textValue: "",
    value: ""
    // With the type defaulting to 'String', typescript would infer a 'FilterFieldStringNode'.
  };
  const errors = [];
  let positionOperators;
  let positionOperatorsInsideString;
  cursor.iterate((childNode) => {
    const { from: childFrom, to: childTo } = childNode;
    const nodeName = childNode.name;
    switch (nodeName) {
      case "FilterSimpleValue":
      case "ListValue":
        break;
      case "Identifier":
      case "String":
      case "IncompleteString":
      case "Variable": {
        const {
          convertedNode: { textValue, value },
          positionOperators: { endsWithOperator, startsWithOperator },
          isValid: isValidIdentifier
        } = convertIdentifier(childNode, doc, nodeName);
        if (
          // eslint-disable-next-line @typescript-eslint/no-deprecated
          parserConfig?.ipAddressConversion && nodeName === "Identifier" && isValidIPAddress(value)
        ) {
          convertedNode.type = "IPAddress";
        } else if (
          // eslint-disable-next-line @typescript-eslint/no-deprecated
          parserConfig?.timestampConversion && (nodeName === "Identifier" || nodeName === "String") && isValidTimestamp(value)
        ) {
          convertedNode.type = "Timestamp";
        }
        if (
          // eslint-disable-next-line @typescript-eslint/no-deprecated
          parserConfig?.uidConversion && nodeName === "Identifier" && isValidUID(value)
        ) {
          convertedNode.type = "UID";
        }
        if (
          // eslint-disable-next-line @typescript-eslint/no-deprecated
          parserConfig?.smartscapeIdConversion && nodeName === "Identifier" && isValidSmartscapeId(value)
        ) {
          convertedNode.type = "SmartscapeId";
        }
        const customType = isValidCustomType(value, customTypes);
        if (customType) {
          convertedNode.type = customType;
        }
        if (endsWithOperator || startsWithOperator) {
          if (!positionOperatorsInsideString) {
            positionOperatorsInsideString = { ranges: [] };
          }
          if (endsWithOperator) {
            positionOperatorsInsideString.type = "ends-with";
            positionOperatorsInsideString.ranges.push(endsWithOperator);
          }
          if (startsWithOperator) {
            positionOperatorsInsideString.type = "starts-with";
            positionOperatorsInsideString.ranges.push(startsWithOperator);
          }
          if (endsWithOperator && startsWithOperator) {
            positionOperatorsInsideString.type = "contains";
            if (endsWithOperator.from === startsWithOperator.from && endsWithOperator.to === startsWithOperator.to) {
              convertToErrorNode(convertedNode);
              errors.push({
                range: positionOperatorsInsideString.ranges[0],
                message: "unsupportedPositionOperator"
              });
            }
          }
        }
        updateNodeValues(convertedNode, textValue, value);
        if (nodeName === "Variable") {
          convertedNode.type = "Variable";
          convertedNode.variableType = "Any";
        }
        if (!isValidIdentifier || nodeName === "IncompleteString") {
          convertToErrorNode(convertedNode);
        }
        break;
      }
      case "InvalidIdentifier": {
        const {
          convertedNode: { textValue }
        } = convertNode(childNode, doc);
        updateNodeValues(convertedNode, textValue);
        convertToErrorNode(convertedNode);
        break;
      }
      case "Boolean": {
        const { convertedNode: booleanNode } = convertBoolean(childNode, doc);
        convertedNode = {
          ...convertedNode,
          ...booleanNode
        };
        break;
      }
      case "Number": {
        ({ convertedNode } = convertNumber(childNode, doc, numberParser));
        break;
      }
      case "Duration":
        ({ convertedNode } = convertDuration(childNode, doc));
        break;
      case "PositionOperator": {
        const text = getTextFromRange(childNode, doc);
        updateNodeValues(convertedNode, text, "");
        if (!positionOperators) {
          positionOperators = { ranges: [] };
        }
        const childRange = { from: childFrom, to: childTo };
        if (childFrom === from) {
          positionOperators.type = "ends-with";
          positionOperators.ranges.push(childRange);
        } else if (childTo === to) {
          positionOperators.type = positionOperators.type ? "contains" : "starts-with";
          positionOperators.ranges.push(childRange);
        }
        break;
      }
      case "InvalidPositionOperator": {
        const text = getTextFromRange(childNode, doc);
        updateNodeValues(convertedNode, text);
        convertToErrorNode(convertedNode);
        if (nodeName === "InvalidPositionOperator") {
          errors.push({
            range: { from: childFrom, to: childTo },
            message: "unsupportedPositionOperator"
          });
        }
        break;
      }
      case "\u26A0": {
        convertedNode.value = convertedNode.textValue;
        convertToErrorNode(convertedNode);
        return false;
      }
      default:
        if (isDev()) {
          console.log(
            `${doc.toString()} | Syntax node '${childNode.name}' is not yet handled in the simple value conversion. Please make sure to handle all nodes accordingly.`
          );
        }
    }
  });
  if (convertedNode.type === "Error") {
    convertedNode.textValue = convertedNode.value = getTextFromRange(
      convertedNode.range,
      doc
    );
    errors.push({
      range: { from, to },
      message: "invalidValueErrorGeneric"
    });
  }
  let combinedPositionOperators = positionOperators ?? positionOperatorsInsideString;
  if (positionOperators && positionOperatorsInsideString) {
    combinedPositionOperators = {
      ...positionOperators,
      ranges: [
        ...positionOperators?.ranges ?? [],
        ...positionOperatorsInsideString?.ranges ?? []
      ]
    };
    const { type } = positionOperators;
    const { type: typeInsideString } = positionOperatorsInsideString;
    if (
      // foo = *"*bar" | foo = *"*bar"*
      typeInsideString === "ends-with" && (type === "ends-with" || type === "contains") || // foo = "bar*"* | foo = *"bar*"*
      typeInsideString === "starts-with" && (type === "starts-with" || type === "contains") || // foo = *"*bar*"*
      typeInsideString === "contains" && type === "contains"
    ) {
      convertToErrorNode(convertedNode);
      errors.push({
        range: positionOperatorsInsideString.ranges[0],
        message: "duplicatePositionOperator"
      });
      errors.push({
        range: positionOperators.ranges[0],
        message: "duplicatePositionOperator"
      });
    }
    if (typeInsideString === "contains" && type === "contains") {
      convertToErrorNode(convertedNode);
      errors.push({
        range: positionOperatorsInsideString.ranges[1],
        message: "duplicatePositionOperator"
      });
      errors.push({
        range: positionOperators.ranges[1],
        message: "duplicatePositionOperator"
      });
    }
    if (typeInsideString === "ends-with" && type === "starts-with" || typeInsideString === "starts-with" && type === "ends-with") {
      combinedPositionOperators.type = "contains";
    }
  }
  return {
    convertedNode,
    errors,
    positionOperators: errors.length === 0 ? combinedPositionOperators : void 0
  };
}
export {
  convertSimpleValue
};
