// @dynatrace/strato-components/esm/filters/filter-field/error-fallback/ErrorFallback.js
import { jsx, jsxs } from "react/jsx-runtime";
import { FormattedMessage } from "react-intl";
import Borders from "@dynatrace/strato-design-tokens/borders";
import { CriticalIcon, ResetIcon } from "@dynatrace/strato-icons";
import { errorFallbackMessages } from "./messages.js";
import { Button } from "../../../buttons/button/Button.js";
import { Container } from "../../../layouts/container/Container.js";
import { Flex } from "../../../layouts/flex/Flex.js";
import { Text } from "../../../typography/text/Text.js";
const ErrorFallback = ({ resetErrorBoundary }) => {
  return /* @__PURE__ */ jsx(
    Container,
    {
      padding: 2,
      paddingLeft: 4,
      color: "critical",
      style: {
        borderRadius: Borders.Radius.Container.Subdued
      },
      children: /* @__PURE__ */ jsxs(
        Flex,
        {
          flexDirection: "row",
          justifyContent: "space-between",
          alignItems: "center",
          gap: 4,
          children: [
            /* @__PURE__ */ jsx(CriticalIcon, {}),
            /* @__PURE__ */ jsx(
              Text,
              {
                textStyle: "base-emphasized",
                style: {
                  flex: 1,
                  overflow: "hidden",
                  whiteSpace: "nowrap",
                  textOverflow: "ellipsis"
                },
                children: /* @__PURE__ */ jsx(FormattedMessage, { ...errorFallbackMessages.errorTitleMessage })
              }
            ),
            /* @__PURE__ */ jsxs(
              Button,
              {
                variant: "emphasized",
                color: "critical",
                onClick: resetErrorBoundary,
                size: "condensed",
                style: { borderRadius: Borders.Radius.Field.Subdued },
                children: [
                  /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(ResetIcon, {}) }),
                  /* @__PURE__ */ jsx(FormattedMessage, { ...errorFallbackMessages.recoverStateButtonText })
                ]
              }
            )
          ]
        }
      )
    }
  );
};
export {
  ErrorFallback
};
