// @dynatrace/strato-components/esm/filters/filter-field/error-fallback/filter-value-obfuscation-util.js
import { convertStringToFilterFieldTree } from "../utils/conversion/string-tree-conversion.js";
import { convertFilterFieldTreeToString } from "../utils/conversion/tree-string-conversion.js";
import { RESERVED_CHARACTERS } from "../utils/escape-suggestion.js";
const TYPE_REPLACEMENT_MAP = {
  Key: "k",
  String: "s",
  JSONPath: "j",
  Variable: "v",
  Number: "n",
  BigInt: "b",
  Duration: "d",
  IPAddress: "i",
  Error: "e",
  UID: "u",
  SmartscapeId: "c",
  Timestamp: "t",
  // The following are not obfuscated
  LogicalOperator: null,
  ComparisonOperator: null,
  ExistsOperator: null,
  SearchOperator: null,
  InclusionOperator: null,
  Boolean: null
};
const EXTENDED_RESERVED_CHARACTERS = `.${RESERVED_CHARACTERS}`;
function getReplacementChar(node) {
  if (node.type in TYPE_REPLACEMENT_MAP) {
    return TYPE_REPLACEMENT_MAP[node.type];
  }
  return "U";
}
function obfuscateText(textValue, replacementChar) {
  if (textValue === void 0) {
    return "";
  }
  return textValue.split("").map((char) => {
    if (EXTENDED_RESERVED_CHARACTERS.includes(char)) {
      return char;
    }
    return replacementChar;
  }).join("");
}
function obfuscateFilterFieldLeafNode(node) {
  if (node === void 0) {
    return node;
  }
  if (node.type === "List") {
    return {
      ...node,
      value: node.value.map((value) => obfuscateFilterFieldLeafNode(value))
    };
  }
  const replacementChar = getReplacementChar(node);
  if (!replacementChar) {
    return node;
  }
  return {
    ...node,
    // Special handling for Duration type to avoid obfuscating reserved characters like 's', 'm', 'h', etc.
    textValue: node.type === "Duration" ? "" : obfuscateText(node.textValue, replacementChar),
    value: obfuscateText(String(node.value), replacementChar)
  };
}
function obfuscateStatement(node) {
  return {
    ...node,
    key: obfuscateFilterFieldLeafNode(node.key),
    comparisonOperator: node.comparisonOperator,
    value: obfuscateFilterFieldLeafNode(node.value)
  };
}
function obfuscateGroupNode(groupNode) {
  return {
    ...groupNode,
    children: groupNode.children.map((node) => {
      if (node.type === "Group") {
        return obfuscateGroupNode(node);
      }
      if (node.type === "Statement") {
        return obfuscateStatement(node);
      }
      return node;
    })
  };
}
function obfuscateFilterString(filterString, validatorMap, parserConfig) {
  const tree = convertStringToFilterFieldTree(filterString, {
    validatorMap,
    parserConfig
  });
  if (tree.tree) {
    return convertFilterFieldTreeToString(obfuscateGroupNode(tree.tree));
  }
  return "no filter tree";
}
export {
  obfuscateFilterString
};
