// @dynatrace/strato-components/core/time/helpers.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var helpers_exports = {};
__export(helpers_exports, {
  getFormattedMinMax: () => getFormattedMinMax,
  getFormattedTimeValue: () => getFormattedTimeValue,
  isEqualRangeSelection: () => isEqualRangeSelection,
  isEqualTimeValue: () => isEqualTimeValue,
  isFullTimeframe: () => isFullTimeframe,
  isPartialTimeframe: () => isPartialTimeframe,
  isSimpleTimeOffset: () => isSimpleTimeOffset,
  isTimeValue: () => isTimeValue,
  isTimeframe: () => isTimeframe,
  mergeRangeValue: () => mergeRangeValue
});
module.exports = __toCommonJS(helpers_exports);
var import_tz = require("@date-fns/tz");
var import_date_fns = require("date-fns");
var import_units = require("@dynatrace-sdk/units");
var import_user_preferences = require("@dynatrace-sdk/user-preferences");
function isEqualTimeValue(tv1, tv2) {
  if (tv1 === tv2) {
    return true;
  }
  if (typeof tv1 === "string" || typeof tv2 === "string") {
    return false;
  }
  if (tv1 === null || tv2 === null) {
    return false;
  }
  if (tv1?.type !== tv2?.type) {
    return false;
  }
  return tv1?.value === tv2?.value;
}
function isEqualRangeSelection(tf1, tf2) {
  if (tf1 === tf2) {
    return true;
  }
  return isEqualTimeValue(tf1?.from ?? null, tf2?.from ?? null) && isEqualTimeValue(tf1?.to ?? null, tf2?.to ?? null);
}
function isSimpleTimeOffset(offset) {
  return typeof offset !== "object";
}
function isTimeframe(date, type) {
  return type === "range";
}
function isTimeValue(value) {
  return typeof value === "object" && value !== null && "absoluteDate" in value && "value" in value && "type" in value;
}
function isFullTimeframe(timeframe) {
  if (typeof timeframe !== "object" || timeframe === null) {
    return false;
  }
  const candidate = timeframe;
  return isTimeValue(candidate.from) && candidate.from.absoluteDate !== "" && isTimeValue(candidate.to) && candidate.to.absoluteDate !== "";
}
function isPartialTimeframe(timeframe) {
  if (typeof timeframe !== "object" || timeframe === null) {
    return false;
  }
  const candidate = timeframe;
  const hasFrom = "from" in candidate && candidate.from !== null;
  const hasTo = "to" in candidate && candidate.to !== null;
  return hasFrom !== hasTo;
}
function getFormattedTimeValue(timeValue, precision, showTimeOnly) {
  if (!timeValue) {
    return "";
  }
  const timezone = (0, import_user_preferences.getTimezone)();
  const formatOptions = {
    // Extract the user-settings timezone to use for the calculation
    timeZone: (0, import_user_preferences.getTimezone)(),
    day: showTimeOnly ? void 0 : "numeric",
    month: showTimeOnly ? void 0 : "short",
    year: showTimeOnly ? void 0 : "numeric",
    hour: "numeric",
    minute: "numeric",
    second: precision === "minutes" ? void 0 : "numeric",
    fractionalSecondDigits: precision === "milliseconds" ? 3 : void 0
  };
  const absoluteDate = timeValue.absoluteDate ? new import_tz.TZDateMini(timeValue.absoluteDate, timezone) : null;
  return absoluteDate && (0, import_date_fns.isValid)(absoluteDate) ? (0, import_units.formatDate)(absoluteDate, formatOptions) : "";
}
function getFormattedMinMax(min, max, precision) {
  const relativeNow = import_tz.TZDateMini.tz((0, import_user_preferences.getTimezone)()).getTime();
  const minDate = min ? (0, import_units.parseTimeAsTimeValue)(min, relativeNow) : null;
  const maxDate = max ? (0, import_units.parseTimeAsTimeValue)(max, relativeNow) : null;
  const newMinDate = new import_tz.TZDateMini(minDate?.absoluteDate ?? "", (0, import_user_preferences.getTimezone)());
  if (!isNaN(newMinDate.getTime())) {
    if (precision === "seconds" && newMinDate.getMilliseconds() > 0) {
      newMinDate.setSeconds(newMinDate.getSeconds() + 1);
    } else if (precision === "minutes" && (newMinDate.getMilliseconds() > 0 || newMinDate.getSeconds() > 0)) {
      newMinDate.setMinutes(newMinDate.getMinutes() + 1);
    }
  }
  const minTimeValue = (0, import_units.parseTimeAsTimeValue)(
    !isNaN(newMinDate.getTime()) ? newMinDate.toISOString() : minDate?.absoluteDate,
    relativeNow
  );
  const formattedMin = getFormattedTimeValue(minTimeValue, precision);
  const formattedMax = getFormattedTimeValue(
    maxDate,
    precision,
    !!minTimeValue && !!maxDate && (0, import_date_fns.isSameDay)(minTimeValue.absoluteDate, maxDate.absoluteDate)
  );
  return {
    min: formattedMin,
    max: formattedMax
  };
}
function mergeRangeValue(current, patch) {
  const base = current ?? { from: null, to: null };
  return {
    from: "from" in patch ? patch.from ?? null : base.from,
    to: "to" in patch ? patch.to ?? null : base.to
  };
}
