// @dynatrace/strato-components/core/time/useDateTimeMessages.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useDateTimeMessages_exports = {};
__export(useDateTimeMessages_exports, {
  getPlaceholder: () => getPlaceholder,
  useDateTimeMessages: () => useDateTimeMessages
});
module.exports = __toCommonJS(useDateTimeMessages_exports);
var import_tz = require("@date-fns/tz");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_user_preferences = require("@dynatrace-sdk/user-preferences");
var import_intl_cache = require("../../forms/date-time-picker/utils/intl-cache.js");
var import_helpers = require("../../forms/forms-core/time/helpers.js");
const placeholderMessages = (0, import_react_intl.defineMessages)({
  year: {
    defaultMessage: "YYYY",
    id: "KgAwef2nXv2+/xgE",
    description: "The abbreviation of the year part of a date value"
  },
  month: {
    defaultMessage: "MM",
    id: "Bo8Z1UaJVL3ONFo8",
    description: "The abbreviation of the month part of a date value"
  },
  day: {
    defaultMessage: "DD",
    id: "42g95tLLPXcqbFLl",
    description: "The abbreviation of the day part of a date value"
  },
  hour: {
    defaultMessage: "hh",
    id: "5+yhrNhI2AlQmDAo",
    description: "The abbreviation of the hour part of a time value"
  },
  minute: {
    defaultMessage: "mm",
    id: "Xup0Ti6gaxC1KDSY",
    description: "The abbreviation of the minute part of a time value"
  },
  second: {
    defaultMessage: "ss",
    id: "ktoTdX660gW5R80Z",
    description: "The abbreviation of the second part of a time value"
  },
  fractionalSecond: {
    defaultMessage: "sss",
    id: "ERAhDK0Wy/n7Sd2e",
    description: "The abbreviation of the fractional of the second part of a time value"
  }
});
function getPlaceholder(type, formatMessage) {
  const messageDescriptor = placeholderMessages[type];
  if (messageDescriptor) {
    return formatMessage(messageDescriptor);
  }
  if (type === "dayPeriod") {
    return (0, import_helpers.getDayPeriodValues)().join("/");
  }
  return type;
}
const useDateTimeMessages = (options) => {
  const {
    locale = (0, import_helpers.getSafeRegionalFormat)(),
    timeZone: timeZoneProp = (0, import_user_preferences.getTimezone)(),
    type = "datetime",
    precision = "minutes"
  } = options ?? {};
  const { formatMessage } = (0, import_react_intl.useIntl)();
  const { formatDateToParts, timeZone } = (0, import_react.useMemo)(
    () => (0, import_react_intl.createIntl)({ locale, timeZone: timeZoneProp }, import_intl_cache.intlCache),
    [locale, timeZoneProp]
  );
  const formatOptions = (0, import_react.useMemo)(() => {
    const options2 = {};
    switch (type) {
      case "date":
        options2["year"] = "numeric";
        options2["month"] = "2-digit";
        options2["day"] = "2-digit";
        return options2;
      // Return config and ignore precision prop if only the date is shown.
      case "datetime":
        options2["year"] = "numeric";
        options2["month"] = "2-digit";
        options2["day"] = "2-digit";
        options2["hour"] = "2-digit";
        options2["minute"] = "2-digit";
        break;
      case "time":
        options2["hour"] = "numeric";
        options2["minute"] = "2-digit";
        break;
    }
    switch (precision) {
      case "seconds":
        options2["second"] = "2-digit";
        break;
      case "milliseconds":
        options2["second"] = "2-digit";
        options2["fractionalSecondDigits"] = 3;
        break;
    }
    return options2;
  }, [type, precision]);
  const parts = (0, import_react.useMemo)(
    () => formatDateToParts(import_tz.TZDateMini.tz((0, import_user_preferences.getTimezone)()).getTime(), formatOptions),
    [formatOptions, formatDateToParts]
  );
  const formatString = (0, import_react.useMemo)(
    () => parts.map(
      ({ type: type2, value }) => type2 === "literal" ? value : getPlaceholder(type2, formatMessage)
    ).join(""),
    [formatMessage, parts]
  );
  return {
    locale,
    timeZone,
    formatDateToParts,
    parts,
    formatOptions,
    formatString,
    getPlaceholder
  };
};
