// @dynatrace/strato-components/core/time/validation.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var validation_exports = {};
__export(validation_exports, {
  getTimeBoundaryMessageData: () => getTimeBoundaryMessageData,
  isOutOfBounds: () => isOutOfBounds,
  validate: () => validate
});
module.exports = __toCommonJS(validation_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_tz = require("@date-fns/tz");
var import_date_fns = require("date-fns");
var import_react_intl = require("react-intl");
var import_units = require("@dynatrace-sdk/units");
var import_user_preferences = require("@dynatrace-sdk/user-preferences");
var import_helpers = require("./helpers.js");
var import_formfield_messages_state_reducer = require("../../forms/forms-core/state/formfield-messages-state-reducer.js");
var import_date_time = require("../../forms/forms-core/time/date-time.js");
function isOutOfBounds(date, min, max) {
  const relativeNow = import_tz.TZDateMini.tz((0, import_user_preferences.getTimezone)()).getTime();
  const newDate = typeof date === "string" ? (0, import_units.parseTimeAsTimeValue)(date, relativeNow)?.absoluteDate : date;
  if (!newDate) {
    return void 0;
  }
  const minDate = min?.type === "expression" ? (0, import_units.parseTimeAsTimeValue)(min.value, relativeNow)?.absoluteDate : min?.absoluteDate;
  const maxDate = max?.type === "expression" ? (0, import_units.parseTimeAsTimeValue)(max.value, relativeNow)?.absoluteDate : max?.absoluteDate;
  if (minDate && maxDate && ((0, import_date_fns.isBefore)(newDate, minDate) || (0, import_date_fns.isAfter)(newDate, maxDate))) {
    return "minMaxBoundary";
  }
  if (minDate && !maxDate && (0, import_date_fns.isBefore)(newDate, minDate)) {
    return "minBoundary";
  }
  if (!minDate && maxDate && (0, import_date_fns.isAfter)(newDate, maxDate)) {
    return "maxBoundary";
  }
  return void 0;
}
const DEFAULT_ERROR_MESSAGES = (0, import_react_intl.defineMessages)({
  minBoundary: {
    id: "KETnWaCwo9qLRQrI",
    defaultMessage: "Only dates after {min} are available.",
    description: "Given dates are not within the valid range."
  },
  maxBoundary: {
    id: "ONYnUtFidz4+6NmE",
    defaultMessage: "Only dates before {max} are available.",
    description: "Given dates are not within the valid range."
  },
  minMaxBoundary: {
    id: "Qm0/7EQKVXjS9tDM",
    defaultMessage: "Only dates between {min} \u2192 {max} are available.",
    description: "Given dates are not within the valid range."
  },
  invalidOrder: {
    defaultMessage: "'from' date should be before 'to' date",
    id: "DKFIWE64kHV1ohuk",
    description: "Message to display if the from and to dates are in the wrong order."
  },
  singleValueMissing: {
    defaultMessage: "Please fill in this field.",
    id: "hIK5810KwQytwP1q",
    description: "DateTimePicker required value is missing."
  },
  missingFrom: {
    defaultMessage: "Please enter a start time.",
    id: "GUeetpTpusqFkMi+",
    description: "From value is missing."
  },
  missingTo: {
    defaultMessage: "Please enter an end time.",
    id: "tfHTZkRc7sr/Ypcl",
    description: "End value is missing."
  },
  missingFromTo: {
    defaultMessage: "Please enter a start and end time.",
    id: "4yANH5gdmW1efk0I",
    description: "Start and end values are missing."
  },
  valueMissing: {
    defaultMessage: "Empty timeframe.",
    id: "QsKlSIxE+Yp58CUG",
    description: "Timeframe selector required timeframe error hint."
  },
  invalid: {
    defaultMessage: "Invalid time. Expected format: {expectedFormat}, -2h or -2d@d",
    id: "r/XpT4mat98jCrup",
    description: "Given value is invalid"
  },
  invalidFrom: {
    defaultMessage: "Invalid start time. Expected format: {expectedFormat}, -2h or -2d@d",
    id: "supuOfI2cSrNKOZ/",
    description: "Given from value is invalid"
  },
  invalidTo: {
    defaultMessage: "Invalid end time. Expected format: {expectedFormat}, -2h or -2d@d",
    id: "HL9dGWGnhYkm7OyC",
    description: "Given to value is invalid"
  },
  invalidFromTo: {
    defaultMessage: "Invalid start and end time. Expected format: {expectedFormat}, -2h or -2d@d",
    id: "89sBAyzLbeW6GyPJ",
    description: "Given from and to values are invalid"
  }
});
function isValueOutsideGrailRange(value) {
  if (!value) {
    return false;
  }
  const grailMinDate = new Date(import_date_time.GRAIL_MIN_DATE);
  const grailMaxDate = new Date(import_date_time.GRAIL_MAX_DATE);
  if ((0, import_helpers.isTimeValue)(value)) {
    if (!value.absoluteDate) {
      return false;
    }
    const date = new Date(value.absoluteDate);
    return (0, import_date_fns.isBefore)(date, grailMinDate) || (0, import_date_fns.isAfter)(date, grailMaxDate);
  }
  const fromDate = value.from?.absoluteDate ? new Date(value.from.absoluteDate) : null;
  const toDate = value.to?.absoluteDate ? new Date(value.to.absoluteDate) : null;
  return fromDate !== null && ((0, import_date_fns.isBefore)(fromDate, grailMinDate) || (0, import_date_fns.isAfter)(fromDate, grailMaxDate)) || toDate !== null && ((0, import_date_fns.isBefore)(toDate, grailMinDate) || (0, import_date_fns.isAfter)(toDate, grailMaxDate));
}
function filterGrailTimeValue(value, limit) {
  if (!limit) {
    return limit;
  }
  const isGrailLimit = limit.value === import_date_time.GRAIL_MIN_DATE || limit.value === import_date_time.GRAIL_MAX_DATE;
  if (!isGrailLimit) {
    return limit;
  }
  return isValueOutsideGrailRange(value) ? limit : null;
}
function isTimeBoundaryErrorKey(errorKey) {
  return errorKey === "minBoundary" || errorKey === "maxBoundary" || errorKey === "minMaxBoundary";
}
function parseLimit(limit, relativeNow = import_tz.TZDateMini.tz((0, import_user_preferences.getTimezone)()).getTime()) {
  if (typeof limit !== "string") {
    return limit;
  }
  return (0, import_units.parseTimeAsTimeValue)(limit, relativeNow) ?? null;
}
function getBoundaryErrorKey(min, max) {
  if (min && max) {
    return "minMaxBoundary";
  }
  if (min) {
    return "minBoundary";
  }
  if (max) {
    return "maxBoundary";
  }
  return void 0;
}
function getTimeBoundaryMessageData(value, min, max, precision = "minutes", onlyWhenOutOfBounds = true) {
  const relativeNow = import_tz.TZDateMini.tz((0, import_user_preferences.getTimezone)()).getTime();
  const parsedMin = parseLimit(min, relativeNow);
  const parsedMax = parseLimit(max, relativeNow);
  const filteredMin = filterGrailTimeValue(value, parsedMin);
  const filteredMax = filterGrailTimeValue(value, parsedMax);
  const errorKey = onlyWhenOutOfBounds ? getOutOfBoundsErrorKey(value, filteredMin, filteredMax) : getBoundaryErrorKey(filteredMin, filteredMax);
  if (!isTimeBoundaryErrorKey(errorKey)) {
    return void 0;
  }
  const { min: formattedMin, max: formattedMax } = (0, import_helpers.getFormattedMinMax)(
    filteredMin?.value,
    filteredMax?.value,
    precision
  );
  return {
    errorKey,
    values: {
      min: formattedMin,
      max: formattedMax
    }
  };
}
function getOutOfBoundsErrorKey(updatedValue, min, max) {
  if (updatedValue && ("from" in updatedValue || "to" in updatedValue) && (0, import_helpers.isTimeframe)(updatedValue, "range")) {
    const fromErrorKey = isOutOfBounds(updatedValue?.from?.value, min, max);
    const toErrorKey = isOutOfBounds(updatedValue?.to?.value, min, max);
    return fromErrorKey || toErrorKey;
  }
  if (updatedValue && (0, import_helpers.isTimeValue)(updatedValue)) {
    return isOutOfBounds(updatedValue.value, min, max);
  }
}
function isInvalidOrder(value) {
  if ((0, import_helpers.isFullTimeframe)(value)) {
    const timezone = (0, import_user_preferences.getTimezone)();
    const fromTime = new import_tz.TZDateMini(
      value.from?.absoluteDate,
      timezone
    ).getTime();
    const toTime = new import_tz.TZDateMini(value.to?.absoluteDate, timezone).getTime();
    if (value.from && value.to && fromTime >= toTime) {
      return true;
    }
  }
  return false;
}
function getInvalidErrorKey(value, valueMode) {
  if (valueMode !== "single" && value && !(0, import_helpers.isTimeValue)(value) && ("from" in value && value.from !== void 0 || "to" in value && value.to !== void 0)) {
    const isFromInvalid = value.from?.absoluteDate === "" && value.from?.value !== "";
    const isToInvalid = value.to?.absoluteDate === "" && value.to?.value !== "";
    if (isFromInvalid && isToInvalid) {
      return "invalidFromTo";
    }
    if (isFromInvalid) {
      return "invalidFrom";
    }
    if (isToInvalid) {
      return "invalidTo";
    }
  }
  if (valueMode === "single" && value && "absoluteDate" in value && value.absoluteDate === "" && value.value !== "") {
    return "invalid";
  }
  return void 0;
}
function getValueMissingErrorKey(value, valueMode) {
  if (value === null) {
    if (valueMode === "single") {
      return "singleValueMissing";
    }
    if (valueMode === "dateTimePicker") {
      return "missingFromTo";
    }
    return "valueMissing";
  }
  if (!(0, import_helpers.isTimeValue)(value) && value.from === null && value.to === null) {
    if (valueMode === "dateTimePicker") {
      return "missingFromTo";
    }
    return "valueMissing";
  }
  if (valueMode !== "single" && (0, import_helpers.isPartialTimeframe)(value)) {
    if (!value.from) {
      return "missingFrom";
    }
    if (!value.to) {
      return "missingTo";
    }
  }
  return void 0;
}
function validate(value, min, max, options) {
  const {
    formMessageDispatch,
    formerError: previousError,
    precision: validationPrecision = "minutes",
    showError: showValidationError,
    checkOrder: shouldCheckOrder,
    ref: validationRef,
    expectedFormat: expectedValidationFormat = "YYYY-MM-DD HH:mm:ss",
    valueMode: validationValueMode,
    required: isRequired,
    errorMessages
  } = options ?? {};
  const timeBoundaryMessageData = getTimeBoundaryMessageData(
    value,
    min,
    max,
    validationPrecision
  );
  const invalidOrderError = shouldCheckOrder && isInvalidOrder(value) ? "invalidOrder" : void 0;
  const invalidError = getInvalidErrorKey(value, validationValueMode);
  const valueMissingError = isRequired ? getValueMissingErrorKey(value, validationValueMode) : void 0;
  const outOfBoundsError = timeBoundaryMessageData?.errorKey;
  const errorKey = invalidOrderError ?? valueMissingError ?? outOfBoundsError ?? invalidError;
  if (errorKey) {
    const messageDef = errorMessages?.[errorKey] ?? DEFAULT_ERROR_MESSAGES[errorKey];
    validationRef?.current && formMessageDispatch?.({
      type: import_formfield_messages_state_reducer.FormFieldMessagesAction.ADDMESSAGE,
      payload: {
        id: errorKey,
        variant: "error",
        messageType: "formcontrol",
        message: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_react_intl.FormattedMessage,
          {
            ...messageDef,
            values: {
              ...timeBoundaryMessageData?.values,
              expectedFormat: expectedValidationFormat
            }
          }
        ),
        validityStateKey: "customError",
        showError: showValidationError
      }
    });
  }
  if (previousError && previousError !== errorKey) {
    validationRef?.current && formMessageDispatch?.({
      type: import_formfield_messages_state_reducer.FormFieldMessagesAction.REMOVEMESSAGE,
      payload: {
        id: previousError,
        messageType: "formcontrol"
      }
    });
  }
  return errorKey;
}
