// @dynatrace/strato-components/charts/categorical-bar/components/legend/CategoricalBarChartLegend.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CategoricalBarChartLegend_exports = {};
__export(CategoricalBarChartLegend_exports, {
  CategoricalBarChartLegend: () => CategoricalBarChartLegend
});
module.exports = __toCommonJS(CategoricalBarChartLegend_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_useChartLayoutLoading = require("../../../core/components/chart-layout/hooks/useChartLayoutLoading.js");
var import_useChartLayoutOrientation = require("../../../core/components/chart-layout/hooks/useChartLayoutOrientation.js");
var import_orientation = require("../../../core/components/chart-layout/utils/orientation.js");
var import_ChartLegend = require("../../../core/components/legend/categorical/ChartLegend.js");
var import_useLegendSeriesActions = require("../../../core/components/legend/categorical/hooks/useLegendSeriesActions.js");
var import_safe_get = require("../../../core/utils/safe-get.js");
var import_selectors = require("../../hooks/store/selectors.js");
var import_useCategoricalBarChartData = require("../../hooks/useCategoricalBarChartData.js");
var import_useTruncationMode = require("../../hooks/useTruncationMode.js");
function getLegendItems(categories, dimensionMap, uniqueDimensions) {
  const dimensionMapCache = /* @__PURE__ */ new Map();
  const uniqueDimensionsSet = new Set(uniqueDimensions);
  categories.forEach((category) => {
    category.dimensions.forEach((dimension) => {
      const { name } = (0, import_safe_get.safeGet)(dimensionMap, dimension);
      if (uniqueDimensionsSet.has(name)) {
        if (!dimensionMapCache.has(name)) {
          dimensionMapCache.set(name, {
            dimension,
            values: []
          });
        }
        dimensionMapCache.get(name).values.push({
          category: category.name,
          value: (0, import_safe_get.safeGet)(dimensionMap, dimension)?.value
        });
      }
    });
  });
  const legendItems = Array.from(dimensionMapCache.entries()).map(
    ([uniqueDimensionKey, { dimension, values }]) => {
      const { id, name, color, legendColor } = (0, import_safe_get.safeGet)(dimensionMap, dimension);
      return {
        ...dimension,
        id,
        name,
        color: legendColor ?? color,
        categoryDimensionValues: values,
        category: void 0,
        value: values.length > 1 ? void 0 : dimension.value
      };
    }
  );
  return legendItems;
}
const CategoricalBarChartLegend = (0, import_react.memo)(() => {
  const {
    uniqueDimensions,
    categoricalData: categories,
    dimensionMap
  } = (0, import_useCategoricalBarChartData.useCategoricalBarChartData)();
  const { orientation: layoutOrientation } = (0, import_useChartLayoutOrientation.useChartLayoutOrientation)();
  const truncationMode = (0, import_useTruncationMode.useTruncationMode)();
  const legendOrientation = (0, import_orientation.getLegendOrientation)(layoutOrientation);
  const isLoading = (0, import_useChartLayoutLoading.useChartLayoutLoading)();
  const legendItems = getLegendItems(
    categories,
    dimensionMap,
    uniqueDimensions
  );
  const setState = (0, import_selectors.useSetState)();
  const onDimensionHover = (item) => {
    setState((prev) => ({
      ...prev,
      highlightedDimension: item
    }));
  };
  const onDimensionClick = (visibleDimension) => {
    const visibleSeriesSet = new Set(visibleDimension.map((s) => s.name));
    setState((prev) => ({
      ...prev,
      filteredDimensions: visibleSeriesSet
    }));
  };
  const legendActions = (0, import_useLegendSeriesActions.useLegendSeriesActions)(
    legendItems,
    (c) => c.id
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ChartLegend.ChartLegend,
    {
      items: legendItems,
      orientation: legendOrientation,
      onHover: onDimensionHover,
      onClick: onDimensionClick,
      onSeriesAction: legendActions,
      truncationMode,
      interactionsDisabled: isLoading
    }
  );
});
