// @dynatrace/strato-components/buttons/run-query-button/RunQueryButton.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var RunQueryButton_exports = {};
__export(RunQueryButton_exports, {
  RunQueryButton: () => RunQueryButton,
  queryStateOptions: () => queryStateOptions
});
module.exports = __toCommonJS(RunQueryButton_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_ProgressCircle = require("../../content/progress/ProgressCircle.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_merge_props = require("../../core/utils/merge-props.js");
var import_Button = require("../button/Button.js");
const COMPONENT_NAME = "RunQueryButton";
const queryStateOptions = [
  "idle",
  "loading",
  "success",
  "error"
];
function getVariant(queryState) {
  switch (queryState) {
    case "idle":
      return "accent";
    case "loading":
    case "success":
    case "error":
      return "emphasized";
  }
}
function getColor(queryState) {
  switch (queryState) {
    case "idle":
      return "primary";
    case "loading":
    case "success":
      return "neutral";
    case "error":
      return "critical";
  }
}
function PrefixIcon({
  type,
  ariaLabel
}) {
  switch (type) {
    case "idle":
    case "success":
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.PlayIcon, {});
    case "loading":
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ProgressCircle.ProgressCircle, { size: "small", "aria-label": ariaLabel });
    case "error":
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.CriticalIcon, {});
  }
}
function ButtonMessage(props) {
  const { type, cancelable } = props;
  switch (type) {
    case "idle":
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_react_intl.FormattedMessage,
        {
          description: "Label for button used to run a query",
          defaultMessage: "Run query",
          id: "CwGi8xbTtV+fdsu5"
        }
      );
    case "success":
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_react_intl.FormattedMessage,
        {
          description: "Label for button used to rerun a query after it was successful",
          defaultMessage: "Run query",
          id: "J9M12SSZNn2dKokI"
        }
      );
    case "loading":
      return cancelable ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_react_intl.FormattedMessage,
        {
          description: "Call to action for cancelable query button",
          defaultMessage: "Cancel query",
          id: "Za5bQfBDnyyur0mw"
        }
      ) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_react_intl.FormattedMessage,
        {
          description: "Label for button indicating that a query is loading",
          defaultMessage: "Running",
          id: "hi3IodOo0ygpihhY"
        }
      );
    case "error":
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_react_intl.FormattedMessage,
        {
          description: "Label for button indicating that an error has occurred while running the query. Click to retry.",
          defaultMessage: "Error (click to retry)",
          id: "gg76gjIy5leuPbMT"
        }
      );
  }
}
const RunQueryButton = /* @__PURE__ */ (0, import_react.forwardRef)((props, ref) => {
  const {
    queryState = "idle",
    disabled = false,
    onClick,
    cancelable = true,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const allowsClickHandler = queryState !== "loading" || queryState === "loading" && cancelable;
  const intl = (0, import_react_intl.useIntl)();
  const ariaLabel = intl.formatMessage({
    description: "Label for button indicating that a query is loading",
    defaultMessage: "Running",
    id: "hi3IodOo0ygpihhY"
  });
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_Button.Button,
    {
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      onClick: allowsClickHandler ? onClick : void 0,
      color: getColor(queryState),
      variant: getVariant(queryState),
      ref,
      disabled,
      "aria-live": "polite",
      "aria-atomic": "true",
      className: consumerClassName,
      style: consumerStyle,
      ...(0, import_merge_props.mergeProps)(restProps, behavioralTrackingProps),
      "data-dt-component": COMPONENT_NAME,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(PrefixIcon, { type: queryState, ariaLabel }) }),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(ButtonMessage, { type: queryState, cancelable })
      ]
    }
  );
});
RunQueryButton.displayName = COMPONENT_NAME;
