// @dynatrace/strato-components/charts/timeseries/components/overlays/utils/tooltip/find-closest-points.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var find_closest_points_exports = {};
__export(find_closest_points_exports, {
  filterOutOfBounds: () => filterOutOfBounds,
  findClosestPoints: () => findClosestPoints,
  getClosestPoint: () => getClosestPoint
});
module.exports = __toCommonJS(find_closest_points_exports);
var import_lodash_es = require("lodash");
var import_constants = require("../../../../../core/components/data-point/constants.js");
var import_safe_get = require("../../../../../core/utils/safe-get.js");
var import_sanitize_zero_values = require("../../../../../core/utils/sanitize-zero-values.js");
var import_binary_search = require("../../../../utils/binary-search.js");
var import_timeseries2 = require("../../../../utils/timeseries.js");
var import_yAxis = require("../../../../utils/yAxis.js");
const getClosestPoint = (xPoints, mouseY = 0) => {
  let aux = Infinity;
  let auxPoint = xPoints[0];
  xPoints.forEach((elem) => {
    if (Math.abs(elem.y - mouseY) < aux) {
      aux = Math.abs(elem.y - mouseY);
      auxPoint = elem;
    }
  });
  return auxPoint;
};
function xAccessor(xAxisScale) {
  return (datapoint) => xAxisScale((0, import_timeseries2.getDatapointTime)(datapoint.start, datapoint.end));
}
function roundTo2Decimals(num) {
  const factor = 100;
  return Math.round(num * factor) / factor;
}
function clusterAccessor(xAxisScale, mouseX) {
  return (datapoint) => roundTo2Decimals(
    Math.abs(
      xAxisScale((0, import_timeseries2.getDatapointTime)(datapoint.start, datapoint.end)) - mouseX
    )
  );
}
function mapToClosestPoints(datapointsMap, xAxisScale, yAxis) {
  return (datapoint) => {
    const datapointMap = (0, import_safe_get.safeGet)(datapointsMap, datapoint);
    const {
      isGap,
      color,
      value,
      unit,
      y0,
      y1,
      cumulativeValue,
      seriesId,
      start,
      end,
      center,
      variant
    } = datapointMap;
    if (isGap) {
      return [];
    }
    const yScale = (0, import_yAxis.getYScaleByUnit)(unit, yAxis);
    const x = xAxisScale((0, import_timeseries2.getDatapointTime)(start, end, center));
    if (variant === "band") {
      const pointY0 = (0, import_sanitize_zero_values.sanitizeZeroValues)(yScale, y0 ?? 0);
      const pointY1 = (0, import_sanitize_zero_values.sanitizeZeroValues)(yScale, y1 ?? 0);
      return [
        { id: seriesId, x, y: pointY0, color, datapoint },
        { id: seriesId, x, y: pointY1, color, datapoint }
      ];
    }
    const y = (0, import_sanitize_zero_values.sanitizeZeroValues)(
      yScale,
      cumulativeValue ? cumulativeValue : value ?? 0
    );
    if (!(0, import_lodash_es.isNaN)(y)) {
      return [{ id: seriesId, x, y, color, datapoint }];
    }
    return [];
  };
}
const findClosestPoints = (mouseX, datapoints, xAxisScale, yAxis, datapointsMap) => {
  const [leftIndex, rightIndex] = (0, import_binary_search.getClusterBoundariesAtPosition)(
    mouseX,
    datapoints,
    xAccessor(xAxisScale),
    clusterAccessor(xAxisScale, mouseX)
  );
  const toClosestPoint = mapToClosestPoints(datapointsMap, xAxisScale, yAxis);
  const closestPoints = [];
  for (let i = leftIndex; i <= rightIndex; i++) {
    closestPoints.push(...toClosestPoint(datapoints[i]));
  }
  return closestPoints;
};
const filterOutOfBounds = (overlaysWidth, overlaysHeight) => (datapoint) => {
  return datapoint.y >= import_constants.OVERLAY_PADDING && datapoint.y <= overlaysHeight + import_constants.OVERLAY_PADDING && datapoint.x >= import_constants.OVERLAY_PADDING && datapoint.x <= overlaysWidth + import_constants.OVERLAY_PADDING;
};
