// @dynatrace/strato-components/charts/timeseries/components/overlays/utils/tooltip/timeseries-tooltip.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var timeseries_tooltip_exports = {};
__export(timeseries_tooltip_exports, {
  filterGroupedPoints: () => filterGroupedPoints,
  filterTimestampsAndTimeframes: () => filterTimestampsAndTimeframes,
  toTooltipItems: () => toTooltipItems
});
module.exports = __toCommonJS(timeseries_tooltip_exports);
var import_lodash_es = require("lodash");
var import_defaultUnitFormatter = require("../../../../../core/utils/defaultUnitFormatter.js");
var import_safe_get = require("../../../../../core/utils/safe-get.js");
const buildBandDataPointValue = (y0, y1, unit, formatter, originalUnit, absoluteY0, absoluteY1) => {
  const [min, max] = [y0, y1].sort((a, b) => a - b);
  const result = [`${formatter(min, unit)} \u2013 ${formatter(max, unit)}`];
  if (absoluteY0 !== void 0 && absoluteY1 !== void 0) {
    const [absoluteMin, absoluteMax] = [absoluteY0, absoluteY1].sort(
      (a, b) => a - b
    );
    result.unshift(
      `${formatter(absoluteMin, originalUnit)} \u2013 ${formatter(
        absoluteMax,
        originalUnit
      )}`
    );
  }
  return result;
};
const buildValuesBasedOnValueRepresentation = (value, unit, originalUnit, formatter, absoluteValue) => {
  const isRelativeRepresentation = absoluteValue !== void 0;
  if (isRelativeRepresentation) {
    return [
      (0, import_defaultUnitFormatter.defaultUnitFormatter)(originalUnit)(absoluteValue),
      formatter(value)
    ];
  }
  return [formatter(value, unit)];
};
const tooltipItemsMemoized = /* @__PURE__ */ new Map();
const toTooltipItems = (xPoints, formatters, datapointsMap, units) => {
  const tooltipItems = [];
  const bandSeries = {};
  const [leftUnit] = units;
  const [leftFormatter, rightFormatter] = formatters;
  xPoints.forEach((xPoint, index) => {
    const xPointId = `${xPoint.id}-${index}-${xPoint.y}-${xPoint.x}`;
    let tooltipItem;
    const memoedTooltipItem = tooltipItemsMemoized.get(xPointId);
    if (memoedTooltipItem) {
      tooltipItem = memoedTooltipItem;
    } else {
      const datapointMetadata = (0, import_safe_get.safeGet)(datapointsMap, xPoint.datapoint);
      const {
        seriesId,
        value,
        absoluteValue,
        y0,
        absoluteY0,
        y1,
        absoluteY1,
        color,
        unit,
        originalUnit,
        seriesName,
        variant
      } = datapointMetadata;
      const formatter = unit === leftUnit ? leftFormatter : rightFormatter;
      const baseTooltipItem = {
        id: `${seriesId}-${index}`,
        name: seriesName,
        color,
        datapoint: xPoint.datapoint
      };
      if (variant === "band") {
        if (!bandSeries[xPoint.id]) {
          tooltipItems.push({
            ...baseTooltipItem,
            asyncValues: () => buildBandDataPointValue(
              // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
              y0,
              // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
              y1,
              unit,
              formatter,
              originalUnit,
              absoluteY0,
              absoluteY1
            )
          });
          bandSeries[xPoint.id] = true;
          return;
        }
        bandSeries[xPoint.id] = false;
        return;
      }
      tooltipItem = {
        ...baseTooltipItem,
        asyncValues: () => buildValuesBasedOnValueRepresentation(
          value ?? 0,
          unit,
          originalUnit ? originalUnit : "",
          formatter,
          absoluteValue
        )
      };
      tooltipItemsMemoized.set(xPointId, tooltipItem);
    }
    tooltipItems.push(tooltipItem);
  });
  return tooltipItems;
};
function filterGroupedPoints(closestPoint, points, groupedPoints) {
  const closestPointGroupIds = groupedPoints.filter((group) => group.map((g) => g.id).includes(closestPoint.id)).flatMap((group) => group.map((point) => point.id));
  return points.filter((point) => closestPointGroupIds.includes(point.id));
}
function filterTimestampsAndTimeframes(points, closestPoint) {
  const filterTimestamps = (point) => (0, import_lodash_es.isUndefined)(point.datapoint.end);
  const filterTimeframes = (point) => !(0, import_lodash_es.isUndefined)(point.datapoint.end);
  const hasTimestampClosestPoint = (0, import_lodash_es.isUndefined)(closestPoint.datapoint.end);
  return hasTimestampClosestPoint ? points.filter(filterTimestamps) : points.filter(filterTimeframes);
}
