// @dynatrace/strato-components/esm/forms/toggle-button-group/Item.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import {
  forwardRef,
  useCallback,
  useEffect,
  useImperativeHandle,
  useRef
} from "react";
import { useToggleButtonGroupContext } from "./hooks/useToggleButtonGroupContext.js";
import { Prefix } from "./Prefix.js";
import * as styles from "./styles.sty.js";
import { Suffix } from "./Suffix.js";
import { VisuallyHidden } from "../../core/components/visually-hidden/VisuallyHidden.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useId } from "../../core/hooks/useId.js";
import { useSlots } from "../../core/hooks/useSlots.js";
import { focusRingStatic } from "../../core/styles/focusRingStatic.sty.js";
import { isDev } from "../../core/utils/is-node-environment.js";
import { mergeProps as _mergeProps } from "../../core/utils/merge-props.js";
import { Flex } from "../../layouts/flex/Flex.js";
import { Tooltip } from "../../overlays/tooltip/Tooltip.js";
import { Text } from "../../typography/text/Text.js";
import { useFormFieldMessagesContext } from "../forms-core/hooks/useFormFieldMessagesContext.js";
import { useFormFieldValidation } from "../forms-core/hooks/useFormFieldValidation.js";
import { FormFieldMessagesAction } from "../forms-core/state/formfield-messages-state-reducer.js";
const Item = forwardRef(
  (props, forwardedRef) => {
    const fallbackId = useId();
    const {
      id = fallbackId,
      disabled: propDisabled,
      readOnly: propReadOnly,
      tooltip,
      children,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      value,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
    const {
      "data-dt-name": dataDtName = "Toggle",
      ...remainingBehavioralTrackingProps
    } = behavioralTrackingProps;
    const {
      name,
      isRequired,
      hasError,
      isDisabled: disabledContext,
      ariaDisabled,
      selectedValue,
      isReadOnly: isReadOnlyContext,
      onChange
    } = useToggleButtonGroupContext();
    const labelRef = useRef(null);
    const inputRef = useRef(null);
    const isDisabled = disabledContext || propDisabled;
    const isReadOnly = isReadOnlyContext || propReadOnly;
    const checked = selectedValue === value;
    const color = hasError ? "critical" : "neutral";
    const { dispatch: messagesContextDispatch } = useFormFieldMessagesContext();
    const {
      onChange: validationOnChange,
      onInvalid: validationOnInvalid,
      onBlur: validationOnBlur,
      ariaProps
    } = useFormFieldValidation(inputRef, { isRequired });
    useEffect(() => {
      messagesContextDispatch?.({
        type: FormFieldMessagesAction.ADDFORMCONTROLID,
        payload: { controlId: id }
      });
      return () => messagesContextDispatch?.({
        type: FormFieldMessagesAction.REMOVEFORMCONTROLID,
        payload: { controlId: id }
      });
    }, [id, messagesContextDispatch]);
    const [slots, freeform] = useSlots(children, {
      prefix: Prefix,
      suffix: Suffix
    });
    useImperativeHandle(forwardedRef, () => {
      return {
        element: labelRef.current,
        inputRef: inputRef.current,
        focus() {
          inputRef.current?.focus();
        },
        validate() {
          inputRef.current?.checkValidity();
        }
      };
    });
    const scrollTabIntoView = useCallback(
      (behavior) => {
        if (!checked || !labelRef.current) {
          return;
        }
        const parent = labelRef.current.parentElement;
        if (!parent) {
          return;
        }
        const parentRect = parent.getBoundingClientRect();
        const targetRect = labelRef.current.getBoundingClientRect();
        const currentLeft = parent.scrollLeft;
        const deltaLeft = targetRect.left - parentRect.left;
        const deltaRight = targetRect.right - parentRect.right;
        let nextLeft = currentLeft;
        if (deltaLeft < 0) {
          nextLeft += deltaLeft;
        } else if (deltaRight > 0) {
          nextLeft += deltaRight;
        }
        parent.scrollTo?.({ left: nextLeft, behavior });
      },
      [checked]
    );
    const handleChange = useCallback(
      (e) => {
        if (isReadOnly) {
          return;
        }
        if (value !== void 0) {
          onChange?.(value, e);
        }
        validationOnChange?.();
      },
      [isReadOnly, value, onChange, validationOnChange]
    );
    useEffect(() => {
      if (checked && labelRef.current) {
        scrollTabIntoView("instant");
      }
    }, []);
    useEffect(() => {
      const parent = labelRef.current?.parentElement;
      if (!parent) {
        return;
      }
      let prevWidth = parent.offsetWidth;
      let debounceTimer;
      const resizeObserver = new window.ResizeObserver((entries) => {
        for (const entry of entries) {
          if (entry.target === parent && parent.offsetWidth !== prevWidth) {
            prevWidth = parent.offsetWidth;
            window.clearTimeout(debounceTimer);
            debounceTimer = window.setTimeout(() => {
              scrollTabIntoView("smooth");
            }, 300);
          }
        }
      });
      resizeObserver.observe(parent);
      scrollTabIntoView("smooth");
      return () => {
        resizeObserver.disconnect();
        window.clearTimeout(debounceTimer);
      };
    }, [checked, scrollTabIntoView]);
    if (isDev() && tooltip === void 0 && freeform.length === 0) {
      console.warn(
        "A text to indicate the value is required. It can be provided by a tooltip."
      );
    }
    const toggleItem = /* @__PURE__ */ jsxs(
      "label",
      {
        "data-testid": dataTestId,
        htmlFor: id,
        ref: labelRef,
        className: clsx(
          styles.item,
          {
            [styles.checked]: checked,
            [styles.disabled]: isDisabled || ariaDisabled,
            [styles.readOnly]: isReadOnly
          },
          focusRingStatic({
            when: "focus-within-visible",
            disabled: isDisabled || ariaDisabled,
            color
          }),
          consumerClassName
        ),
        style: consumerStyle,
        "data-dt-name": dataDtName,
        ...remainingBehavioralTrackingProps,
        ...restProps,
        children: [
          /* @__PURE__ */ jsx(VisuallyHidden, { children: /* @__PURE__ */ jsx(
            "input",
            {
              id,
              type: "radio",
              ref: inputRef,
              name,
              value,
              disabled: isDisabled,
              "aria-disabled": ariaDisabled,
              readOnly: isReadOnly,
              "aria-label": tooltip,
              checked,
              onChange: handleChange,
              required: isRequired,
              onInvalid: validationOnInvalid,
              ..._mergeProps(restProps, {
                ...ariaProps,
                onBlur: validationOnBlur
              })
            }
          ) }),
          slots.prefix,
          freeform.length > 0 && /* @__PURE__ */ jsxs(Flex, { flexDirection: "column", gap: 0, className: styles.textContainer, children: [
            /* @__PURE__ */ jsx(
              Text,
              {
                as: "span",
                textStyle: checked ? "base-emphasized" : "base",
                className: styles.text,
                children: freeform
              }
            ),
            /* @__PURE__ */ jsx(
              Text,
              {
                as: "span",
                textStyle: "base-emphasized",
                className: clsx(styles.hidden, styles.text),
                "aria-hidden": true,
                children: freeform
              }
            )
          ] }),
          slots.suffix
        ]
      }
    );
    return tooltip ? /* @__PURE__ */ jsx(Tooltip, { text: tooltip, children: toggleItem }) : toggleItem;
  }
);
Item.displayName = "ToggleButtonGroup.Item";
export {
  Item
};
