// @dynatrace/strato-components/esm/forms/toggle-button-group/ToggleButtonGroup.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import {
  forwardRef,
  Fragment,
  useCallback,
  useEffect,
  useImperativeHandle,
  useMemo,
  useRef,
  useState
} from "react";
import { ChevronLeftIcon, ChevronRightIcon } from "@dynatrace/strato-icons";
import { ToggleButtonGroupContext } from "./contexts/ToggleButtonGroupContext.js";
import { Item } from "./Item.js";
import { Prefix } from "./Prefix.js";
import * as styles from "./styles.sty.js";
import { Suffix } from "./Suffix.js";
import { Button } from "../../buttons/button/Button.js";
import { useContainerSize } from "../../core/hooks/useContainerSize.js";
import { useControllableState } from "../../core/hooks/useControllableState.js";
import { useId } from "../../core/hooks/useId.js";
import { useSlots } from "../../core/hooks/useSlots.js";
import { isVisualRefreshEnabled } from "../../core/utils/is-visual-refresh-enabled.js";
import { Divider } from "../../layouts/divider/Divider.js";
import { Flex } from "../../layouts/flex/Flex.js";
import { useFormFieldMessagesContext } from "../forms-core/hooks/useFormFieldMessagesContext.js";
import { useSyncedFormControlProps } from "../forms-core/hooks/useSyncedFormControlProps.js";
import { FormFieldMessagesAction } from "../forms-core/state/formfield-messages-state-reducer.js";
const COMPONENT_NAME = "ToggleButtonGroup";
const ToggleButtonGroupRoot = forwardRef((props, forwardedRef) => {
  const fallbackName = useId();
  const {
    children,
    id: propId,
    disabled: propDisabled,
    required: propRequired,
    readOnly: propReadOnly,
    value,
    defaultValue,
    name = fallbackName,
    width = "content",
    onChange,
    className: consumerClassName,
    style: consumerStyle,
    "aria-disabled": propAriaDisabled,
    "data-dtrum-allow": dataDtrumAllow,
    "data-dtrum-mask": dataDtrumMask,
    ...remainingProps
  } = props;
  const wrapperRef = useRef(null);
  const containerRef = useRef(null);
  const [isOverflowingLeft, setIsOverflowingLeft] = useState(false);
  const [isOverflowingRight, setIsOverflowingRight] = useState(false);
  const {
    disabled: formControlDisabled,
    ariaDisabled,
    required: isRequired,
    id: formControlId,
    readOnly: isReadOnly
  } = useSyncedFormControlProps({
    propDisabled,
    propRequired,
    propId,
    propAriaDisabled,
    propReadOnly
  });
  const isDisabled = propDisabled ?? formControlDisabled;
  const isDisabledOrAriaDisabled = isDisabled || ariaDisabled;
  const [selectedValue, setSelectedValue] = useControllableState({
    defaultValue,
    value
  });
  const { dispatch: messagesContextDispatch, state: messageContextState } = useFormFieldMessagesContext();
  const hasError = messageContextState.showError;
  useImperativeHandle(
    forwardedRef,
    () => {
      const newRef = {
        element: wrapperRef.current,
        focus() {
          const checkedInput = wrapperRef.current?.querySelector(
            "input:checked"
          );
          (checkedInput ?? wrapperRef.current?.querySelector("input"))?.focus();
        },
        validate() {
          const inputs = wrapperRef.current?.querySelectorAll("input");
          inputs?.forEach((input) => input?.checkValidity());
          messagesContextDispatch?.({
            type: FormFieldMessagesAction.SETISTOUCHED,
            payload: { isTouched: true }
          });
        }
      };
      if (!wrapperRef.current) {
        return null;
      }
      return new Proxy(wrapperRef.current, {
        get: function(target, property) {
          if (property in newRef) {
            return newRef[property];
          }
          return target[property];
        }
      });
    },
    [messagesContextDispatch]
  );
  const handleChange = useCallback(
    (value2, event) => {
      setSelectedValue?.(value2);
      onChange?.(value2, event);
    },
    [onChange, setSelectedValue]
  );
  const customWidth = width !== "content" && width !== "full" ? width : void 0;
  const [slots] = useSlots(children, {
    items: { match: Item, multiple: true },
    prefix: Prefix,
    suffix: Suffix
  });
  const valueProps = useMemo(
    () => ({
      name,
      isDisabled,
      ariaDisabled,
      isReadOnly,
      isRequired,
      hasError,
      selectedValue,
      onChange: handleChange
    }),
    [
      name,
      isDisabled,
      ariaDisabled,
      isReadOnly,
      isRequired,
      hasError,
      selectedValue,
      handleChange
    ]
  );
  const scrollButtonVariant = isVisualRefreshEnabled() ? "default" : "emphasized";
  const color = hasError && !isDisabledOrAriaDisabled ? "critical" : "neutral";
  const { width: containerWidth } = useContainerSize(containerRef);
  const updateOverflow = useCallback(() => {
    if (containerRef.current) {
      const { scrollLeft, scrollWidth, offsetWidth } = containerRef.current;
      const tolerance = 4;
      setIsOverflowingLeft(scrollLeft > tolerance);
      setIsOverflowingRight(scrollLeft + offsetWidth < scrollWidth - tolerance);
    }
  }, []);
  const slideTabs = (modifier) => {
    const container = containerRef.current;
    if (container) {
      container.scrollBy({
        left: modifier * container.offsetWidth / 2,
        behavior: "smooth"
      });
    }
  };
  useEffect(() => {
    updateOverflow();
  }, [containerWidth, updateOverflow]);
  return /* @__PURE__ */ jsxs(
    "div",
    {
      className: clsx(
        styles.container,
        {
          [styles.disabledGroup]: isDisabledOrAriaDisabled,
          [styles.error]: hasError,
          [styles.readOnly]: isReadOnly,
          [styles.containerWidths.content]: width === "content",
          [styles.containerWidths.full]: width === "full"
        },
        consumerClassName
      ),
      style: { maxWidth: customWidth, ...consumerStyle },
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "aria-orientation": "horizontal",
      role: "radiogroup",
      "aria-live": "polite",
      "aria-disabled": ariaDisabled,
      id: formControlId,
      ref: wrapperRef,
      ...remainingProps,
      "data-dt-component": COMPONENT_NAME,
      children: [
        isOverflowingLeft && /* @__PURE__ */ jsx(
          Button,
          {
            className: styles.scrollButton,
            style: { left: "4px", position: "absolute", top: "4px" },
            variant: scrollButtonVariant,
            color,
            onClick: () => slideTabs(-1),
            size: "condensed",
            children: /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(ChevronLeftIcon, {}) })
          }
        ),
        /* @__PURE__ */ jsx(
          Flex,
          {
            ref: containerRef,
            gap: 0,
            onScroll: updateOverflow,
            className: clsx(styles.scrollContainer, {
              [styles.scrollVariants.left]: isOverflowingLeft && !isOverflowingRight,
              [styles.scrollVariants.right]: !isOverflowingLeft && isOverflowingRight,
              [styles.scrollVariants.both]: isOverflowingLeft && isOverflowingRight
            }),
            children: /* @__PURE__ */ jsx(ToggleButtonGroupContext.Provider, { value: valueProps, children: slots.items?.map((slot, index) => {
              const items = slots.items ?? [];
              const value2 = slot.props.value;
              const isSelected = value2 === selectedValue;
              const isLast = index === items.length - 1;
              const nextItem = items[index + 1];
              const nextItemValue = nextItem?.props?.value;
              const nextIsSelected = nextItemValue === selectedValue;
              return /* @__PURE__ */ jsxs(Fragment, { children: [
                slot,
                !isLast && /* @__PURE__ */ jsx(
                  Divider,
                  {
                    orientation: "vertical",
                    className: clsx(styles.divider, {
                      [styles.hidden]: isSelected || nextIsSelected
                    })
                  }
                )
              ] }, slot.props.value ?? slot.props.id ?? index);
            }) })
          }
        ),
        isOverflowingRight && /* @__PURE__ */ jsx(
          Button,
          {
            className: styles.scrollButton,
            style: { right: "4px", position: "absolute", top: "4px " },
            variant: scrollButtonVariant,
            color,
            onClick: () => slideTabs(1),
            size: "condensed",
            children: /* @__PURE__ */ jsx(Button.Suffix, { children: /* @__PURE__ */ jsx(ChevronRightIcon, {}) })
          }
        )
      ]
    }
  );
});
const ToggleButtonGroup = Object.assign(ToggleButtonGroupRoot, {
  Item,
  Suffix,
  Prefix
});
export {
  ToggleButtonGroup
};
