// @dynatrace/strato-components/charts/histogram/components/overlays/HistogramExplore/HistogramCrosshair.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var HistogramCrosshair_exports = {};
__export(HistogramCrosshair_exports, {
  HistogramCrosshair: () => HistogramCrosshair
});
module.exports = __toCommonJS(HistogramCrosshair_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_useGridPlotSize = require("../../../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_HorizontalCrosshair = require("../../../../core/components/crosshair/HorizontalCrosshair.js");
var import_get_magnifier_label = require("../../../../core/components/crosshair/utils/get-magnifier-label.js");
var import_constants = require("../../../../core/components/data-point/constants.js");
var import_reset_reducers = require("../../../../core/components/interactions-state-helpers/reducers/reset-reducers.js");
var import_constants2 = require("../../../../core/components/threshold-bar/types/constants.js");
var import_useApplyCallbackOnScroll = require("../../../../core/hooks/useApplyCallbackOnScroll.js");
var import_selectors = require("../../../hooks/store/selectors.js");
var import_state_setters = require("../../../hooks/store/state-setters.js");
var import_useHistogramChartOverlayRef = require("../../../hooks/useHistogramChartOverlayRef.js");
var import_useHistogramFilteredData = require("../../../hooks/useHistogramFilteredData.js");
var import_useHistogramThresholds = require("../../../hooks/useHistogramThresholds.js");
var import_useHistogramYFormatter = require("../../../hooks/useHistogramYFormatter.js");
var import_useHistogramYScales = require("../../../hooks/useHistogramYScales.js");
var import_HistogramOverlays_css = require("../HistogramOverlays.sty.js");
const HistogramCrosshair = () => {
  const {
    units: [leftUnit, rightUnit]
  } = (0, import_useHistogramFilteredData.useHistogramFilteredData)();
  const [leftFormatter, rightFormatter] = (0, import_useHistogramYFormatter.useHistogramYFormatter)();
  const { pinned, position } = (0, import_selectors.useExploreState)();
  const setState = (0, import_state_setters.useSetState)();
  const onScrollCallback = () => setState(import_reset_reducers.reduceResetExploreVisible);
  (0, import_useApplyCallbackOnScroll.useApplyCallbackOnScroll)(pinned, onScrollCallback);
  const relativeMousePosition = position?.relative;
  const [leftScale, rightScale] = (0, import_useHistogramYScales.useHistogramYScales)();
  const { width } = (0, import_useGridPlotSize.useGridPlotSize)();
  const { leftThreshold, rightThreshold } = (0, import_useHistogramThresholds.useHistogramThresholds)();
  const closestBinsState = (0, import_selectors.useClosestBinsState)();
  const chartRef = (0, import_useHistogramChartOverlayRef.useHistogramChartOverlayRef)();
  const chartBounds = chartRef.current?.getBoundingClientRect();
  const leftLabel = pinned ? void 0 : (0, import_get_magnifier_label.getMagnifierLabel)(
    leftUnit,
    "y",
    leftFormatter,
    relativeMousePosition,
    leftScale
  );
  const rightLabel = pinned ? void 0 : (0, import_get_magnifier_label.getMagnifierLabel)(
    rightUnit,
    "y",
    rightFormatter,
    relativeMousePosition,
    rightScale
  );
  const hasPinnedBin = pinned && !(0, import_lodash_es.isUndefined)(closestBinsState?.closestBin);
  const canHover = !pinned && !(0, import_lodash_es.isUndefined)(relativeMousePosition);
  if (!hasPinnedBin && !canHover) {
    return null;
  }
  const y = hasPinnedBin ? (
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    closestBinsState.closestBin.y - import_constants.OVERLAY_PADDING
  ) : relativeMousePosition?.y ?? 0;
  const hasLeftThresholds = leftThreshold && leftThreshold.length > 0;
  const hasRightThresholds = rightThreshold && rightThreshold.length > 0;
  const thresholdLeftSize = hasLeftThresholds ? import_constants2.THRESHOLD_BAR_THICKNESS : 0;
  const thresholdRightSize = hasRightThresholds ? import_constants2.THRESHOLD_BAR_THICKNESS : 0;
  const horizontalCrosshairWidth = width + thresholdLeftSize + thresholdRightSize;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("svg", { className: import_HistogramOverlays_css.crosshairLayout, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_HorizontalCrosshair.HorizontalCrosshair,
    {
      "data-testid": "horizontal-crosshair-histogram",
      width: horizontalCrosshairWidth,
      leftLabel,
      rightLabel,
      leftOffset: thresholdLeftSize,
      rightOffset: thresholdRightSize,
      y,
      dashed: !pinned,
      chartBounds
    }
  ) });
};
