// @dynatrace/strato-components/charts/histogram/components/overlays/HistogramExplore/HistogramTooltip.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var HistogramTooltip_exports = {};
__export(HistogramTooltip_exports, {
  HistogramTooltip: () => HistogramTooltip
});
module.exports = __toCommonJS(HistogramTooltip_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_intl = require("react-intl");
var import_HistogramTooltip_css = require("./HistogramTooltip.sty.js");
var import_HistogramTooltipWrapper = require("./HistogramTooltipWrapper.js");
var import_Divider = require("../../../../../layouts/divider/Divider.js");
var import_ChartActionsMenu = require("../../../../core/components/series-action/ChartActionsMenu.js");
var import_useChartActionsMenu = require("../../../../core/components/series-action/hooks/useChartActionsMenu.js");
var import_useDefaultAction = require("../../../../core/components/series-action/hooks/useDefaultAction.js");
var import_InteractiveChartTooltip = require("../../../../core/components/tooltip/InteractiveChartTooltip/components/InteractiveChartTooltip.js");
var import_SeriesBasedContent = require("../../../../core/components/tooltip/InteractiveChartTooltip/contents/SeriesBasedContent.js");
var import_CategoricalTooltipHeader = require("../../../../core/components/tooltip/InteractiveChartTooltip/headers/CategoricalTooltipHeader.js");
var import_useTooltipSeriesActions = require("../../../../core/components/tooltip/shared/hooks/useTooltipSeriesActions.js");
var import_useHistogramFilteredData = require("../../../hooks/useHistogramFilteredData.js");
var import_useHistogramTooltip = require("../../../hooks/useHistogramTooltip.js");
var import_useHistogramTruncationMode = require("../../../hooks/useHistogramTruncationMode.js");
var import_useHistogramXFormatter = require("../../../hooks/useHistogramXFormatter.js");
var import_useHistogramYFormatter = require("../../../hooks/useHistogramYFormatter.js");
var import_useSeriesActionClickHandler = require("../../../hooks/useSeriesActionClickHandler.js");
var import_useXAxisConfig = require("../../../hooks/useXAxisConfig.js");
var import_HistogramDataProvider = require("../../../providers/HistogramDataProvider.js");
var import_constants = require("../constants.js");
var import_bins_to_tooltip = require("../utils/bins-to-tooltip.js");
var import_get_header_label_tooltip = require("../utils/get-header-label-tooltip.js");
function getHeaderResolution(binsMap, bin, formatter) {
  const { from, to } = (0, import_HistogramDataProvider.safeGet)(binsMap, bin);
  return formatter(to - from);
}
const HistogramTooltip = ({
  points,
  closestPoint
}) => {
  const { binsMap, series, seriesMap } = (0, import_useHistogramFilteredData.useHistogramFilteredData)();
  const { unit: xUnit } = (0, import_useXAxisConfig.useXAxisConfig)();
  const { variant } = (0, import_useHistogramTooltip.useHistogramTooltip)();
  const truncationMode = (0, import_useHistogramTruncationMode.useHistogramTruncationMode)();
  const bottomFormatter = (0, import_useHistogramXFormatter.useHistogramXFormatter)();
  const {
    onLeave: handleMenuLeave,
    onScroll: handleMenuScroll,
    selectedItem,
    updateSelectedItem,
    actionsMenuRef,
    isScrolling
  } = (0, import_useChartActionsMenu.useChartActionsMenu)();
  const formatters = (0, import_useHistogramYFormatter.useHistogramYFormatter)();
  const { units } = (0, import_useHistogramFilteredData.useHistogramFilteredData)();
  const isSingle = variant === "single";
  const { tooltipItems, closestTooltipItem } = (0, import_bins_to_tooltip.toTooltipItems)(
    binsMap,
    formatters,
    isSingle,
    units,
    points,
    closestPoint
  );
  const actionClickHandler = (0, import_useSeriesActionClickHandler.useSeriesActionClickHandler)(
    tooltipItems.length > 1
  );
  const seriesActions = (0, import_useTooltipSeriesActions.useTooltipSeriesActions)(
    series,
    (series2) => (0, import_HistogramDataProvider.safeGet)(seriesMap, series2).id,
    tooltipItems.find((t) => t.id === selectedItem)?.bin
  );
  const intl = (0, import_react_intl.useIntl)();
  const showHeaderTooltipItem = !isSingle && tooltipItems.length > 1;
  const showHeaderDivider = tooltipItems.length > 1 && variant === "shared";
  const labelFormatter = (value) => bottomFormatter(value, xUnit);
  const headerLabel = closestPoint ? (0, import_get_header_label_tooltip.getHeaderLabel)(
    binsMap,
    closestPoint.bin,
    labelFormatter,
    intl.formatMessage(import_constants.PREPOSITION.to)
  ) : "";
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_HistogramTooltipWrapper.HistogramTooltipWrapper,
    {
      tooltipItems,
      onMouseLeave: handleMenuLeave,
      onScrollContent: handleMenuScroll,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_InteractiveChartTooltip.InteractiveChartTooltip.Header, { children: closestPoint && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_CategoricalTooltipHeader.CategoricalTooltipHeader,
            {
              resolution: getHeaderResolution(
                binsMap,
                closestPoint.bin,
                labelFormatter
              ),
              name: headerLabel,
              item: showHeaderTooltipItem ? {
                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                ...closestTooltipItem,
                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                values: closestTooltipItem.asyncValues()
              } : void 0
            }
          ),
          showHeaderDivider && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "div",
            {
              className: import_HistogramTooltip_css.wrapper,
              children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Divider.Divider, {})
            },
            `divider-${closestTooltipItem?.id}`
          )
        ] }) }),
        tooltipItems && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_InteractiveChartTooltip.InteractiveChartTooltip.Body, { children: tooltipItems.map((item) => () => {
          const isOpen = selectedItem === item.id;
          return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_ChartActionsMenu.ChartActionsMenu,
            {
              isOpen,
              actions: isOpen ? seriesActions(item) : void 0,
              defaultActions: (0, import_useDefaultAction.getDefaultAction)(item),
              ref: actionsMenuRef,
              onActionClick: actionClickHandler,
              children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_SeriesBasedContent.SeriesBasedContent,
                {
                  item: { ...item, values: item.asyncValues() },
                  onMouseHover: () => updateSelectedItem(item.id),
                  truncationMode,
                  isScrolling
                }
              )
            },
            `${item.id}`
          );
        }) })
      ]
    }
  );
};
