// @dynatrace/strato-components/charts/timeseries/components/axes/TimeseriesTimeBottomAxisArea/TimeseriesTimeBottomAxisArea.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TimeseriesTimeBottomAxisArea_exports = {};
__export(TimeseriesTimeBottomAxisArea_exports, {
  TimeseriesTimeBottomAxisArea: () => TimeseriesTimeBottomAxisArea
});
module.exports = __toCommonJS(TimeseriesTimeBottomAxisArea_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("@use-gesture/react");
var import_clsx = __toESM(require("clsx"));
var import_react2 = require("react");
var import_TimeseriesTimeBottomAxisArea_css = require("./TimeseriesTimeBottomAxisArea.sty.js");
var import_TimeBottomAxis = require("../../../../core/components/axes/components/TimeBottomAxis.js");
var import_ChartGridRoot = require("../../../../core/components/chart-grid/ChartGridRoot.js");
var import_constants = require("../../../../core/components/zoom/constants.js");
var import_ZoomCursors_css = require("../../../../core/components/zoom/ZoomCursors.sty.js");
var import_is_touchable_device = require("../../../../core/utils/is-touchable-device.js");
var import_constants2 = require("../../../constants.js");
var import_selectors = require("../../../hooks/store/selectors.js");
var import_useChartInteractions = require("../../../hooks/useChartInteractions.js");
var import_useTimeseriesTickAxis = require("../../../hooks/useTimeseriesTickAxis.js");
var import_useXAxisConfig = require("../../../hooks/useXAxisConfig.js");
var import_useXScale = require("../../../hooks/useXScale.js");
var import_timeseries_grid_area_rules = require("../../../types/timeseries-grid-area-rules.js");
var import_useApplyPanDelta = require("../../overlays/hooks/useApplyPanDelta.js");
var import_TimeseriesPanningControls = require("../TimeseriesPanningControls.js");
function TimeseriesTimeBottomAxisArea() {
  const xAxisConfig = (0, import_useXAxisConfig.useXAxisConfig)();
  const { currentScale: xScale } = (0, import_useXScale.useXScales)();
  const { pan } = (0, import_useChartInteractions.useChartInteractions)();
  const { zoomOutEnabled } = (0, import_selectors.useZoomState)();
  const isPanEnabled = pan?.enabled && zoomOutEnabled;
  const tickAxis = (0, import_useTimeseriesTickAxis.useTimeseriesTickAxis)();
  const bottomTickAxis = tickAxis?.bottom || { ticks: [], formattedTicks: [] };
  const applyPan = (0, import_useApplyPanDelta.useApplyPanDelta)();
  const axisRef = (0, import_react2.useRef)(null);
  (0, import_react.useGesture)(
    {
      onDrag: (e) => !import_constants.KEYBOARD_HANDLERS.includes(e.type) && applyPan(-e.delta[0])
    },
    {
      target: axisRef,
      eventOptions: { passive: false },
      enabled: isPanEnabled,
      drag: {
        filterTaps: !(0, import_is_touchable_device.isTouchableDevice)(),
        // Without capture:false, tick DOM replacement during panning fires lostpointercapture
        // on the removed element (non-bubbling), leaving useGesture stuck in active drag state.
        pointer: { buttons: [1, 4], capture: false }
      }
    }
  );
  const style = (0, import_is_touchable_device.isTouchableDevice)() ? { touchAction: "none" } : {};
  if (!bottomTickAxis.ticks.length || !xScale) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ChartGridRoot.ChartGrid.GridArea, { gridArea: import_timeseries_grid_area_rules.AREAS.bottomAxis, style, children: ({ availableWidth, availableHeight }) => /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    "div",
    {
      ref: axisRef,
      className: (0, import_clsx.default)(
        (0, import_ZoomCursors_css.cursorStyle)({ isPanEnabled }),
        (0, import_TimeseriesTimeBottomAxisArea_css.timeBottomAxisContainer)({ isTouchableDevice: (0, import_is_touchable_device.isTouchableDevice)() })
      ),
      role: "group",
      "aria-label": import_constants2.TIMESERIES_TIME_AXIS_ARIA_LABEL,
      children: [
        isPanEnabled && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TimeseriesPanningControls.TimeseriesPanningControls, { onClickPan: (d) => applyPan(-d) }),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_TimeBottomAxis.TimeBottomAxis,
          {
            ticks: bottomTickAxis.ticks,
            formattedTicks: bottomTickAxis.formattedTicks,
            scale: xScale,
            containerHeight: availableHeight,
            containerWidth: availableWidth,
            label: xAxisConfig?.label,
            shrinkTickEdges: isPanEnabled
          }
        )
      ]
    }
  ) });
}
TimeseriesTimeBottomAxisArea["displayName"] = "TimeseriesValueBottomAxisArea";
