// @dynatrace/strato-components/esm/core/dt-date/DtDateTime.js
import { Temporal } from "temporal-polyfill";
import { getTimezone } from "@dynatrace-sdk/user-preferences";
class DtDateTime {
  dateTime;
  constructor(dateTimeString, options) {
    const normalized = DtDateTime.normalizeRfc3339(dateTimeString);
    const defaultZone = options?.timezone || getTimezone() || "UTC";
    let parsedZdt = null;
    try {
      parsedZdt = Temporal.Instant.from(normalized);
    } catch {
      parsedZdt = null;
    }
    if (parsedZdt) {
      this.dateTime = parsedZdt.toZonedDateTimeISO(defaultZone);
      return;
    }
    let pdt;
    try {
      pdt = Temporal.PlainDateTime.from(normalized);
    } catch {
      throw new Error(`Invalid datetime string: ${dateTimeString}`);
    }
    this.dateTime = pdt.toZonedDateTime(defaultZone);
  }
  // --- Normalization --------------------------------------------------------
  /**
   * Normalize common RFC3339 variants:
   *  - lowercase 't' → 'T'
   *  - lowercase 'z' → 'Z'
   *  - space between date/time or underscore for Grail support → 'T'
   *
   * We only normalize the first separator between date and time (YYYY-MM-DD <sep> HH:mm…)
   * to avoid mangling valid strings that contain spaces elsewhere (e.g. annotations).
   * See: https://ijmacd.github.io/rfc3339-iso8601/ and https://www.rfc-editor.org/rfc/rfc3339.html#section-5.6
   */
  static normalizeRfc3339(input) {
    let s = input.trim();
    s = s.replace(/^\d{4}-\d{2}-\d{2}[ _]/, (m) => m.replace(/[ _]/, "T"));
    s = s.replace(
      /t/,
      (match, offset) => offset === s.indexOf("t") ? "T" : match
    );
    return s;
  }
  // --- Arithmetic (DST safe via ZonedDateTime) ------------------------------
  addDuration(duration) {
    const next = this.dateTime.add(duration);
    return DtDateTime.fromZdt(next);
  }
  addDays(days) {
    return this.addDuration({ days });
  }
  addWeeks(weeks) {
    return this.addDuration({ weeks });
  }
  // --- Output ---------------------------------------------------------------
  toString() {
    return this.dateTime.toString();
  }
  static fromZdt(zdt) {
    const obj = Object.create(DtDateTime.prototype);
    obj.dateTime = zdt;
    return obj;
  }
}
export {
  DtDateTime
};
