// @dynatrace/strato-components/esm/charts/gantt-chart/components/GanttSegment.js
import { jsx } from "react/jsx-runtime";
import { useContext, useMemo } from "react";
import {
  DEFAULT_BAR_OPACITY,
  STRIPES_PATTERN_ID,
  TIMESTAMP_WIDTH
} from "../constants/defaults.js";
import { GanttSegmentsContext } from "../context/GanttSegmentsContext.js";
import { getEntryColor } from "../functions/get-entry-color.js";
import { useGanttEntryHandlers } from "../hooks/use-gantt-entry-handlers.js";
import { useGanttSegmentPosition } from "../hooks/use-gantt-segment-position.js";
const GanttSegment = (props) => {
  const { start, end, height, containerWidth, data } = props;
  const {
    segmentsData,
    parentSegmentsData,
    config: { colorAccessor, colorPalette }
  } = useContext(GanttSegmentsContext);
  const fill = data ? getEntryColor(data, colorAccessor, colorPalette) : `url(#${STRIPES_PATTERN_ID})`;
  const dataForEntryHandlers = useMemo(
    () => data ? {
      segment: data,
      row: segmentsData,
      parent: parentSegmentsData
    } : void 0,
    [data, segmentsData, parentSegmentsData]
  );
  const {
    onMouseClickHandler,
    onMouseEnterHandler,
    onMouseLeaveHandler,
    onMouseMoveHandler
  } = useGanttEntryHandlers(dataForEntryHandlers);
  if (!isFinite(start) || containerWidth < 0 || end !== void 0 && !isFinite(end)) {
    throw new Error("Invalid input data");
  }
  const {
    start: segmentXStart,
    end: segmentXEnd,
    width: segmentWidth,
    isNegative
  } = useGanttSegmentPosition(start, end ?? start);
  const isTimestamp = end === void 0 || start === end || segmentWidth < TIMESTAMP_WIDTH;
  const segmentX = isNegative ? segmentXEnd : segmentXStart;
  const x = isTimestamp ? segmentX - 1 : segmentX;
  const width = isTimestamp ? TIMESTAMP_WIDTH : segmentWidth;
  return /* @__PURE__ */ jsx(
    "rect",
    {
      onMouseEnter: onMouseEnterHandler,
      onClick: onMouseClickHandler,
      onMouseLeave: onMouseLeaveHandler,
      onMouseMove: onMouseMoveHandler,
      x,
      y: 0,
      opacity: DEFAULT_BAR_OPACITY,
      width,
      height,
      fill,
      pointerEvents: "auto"
    }
  );
};
export {
  GanttSegment
};
