// @dynatrace/strato-components/esm/charts/gantt-chart/components/GanttChartLineRenderer.js
import { jsx, jsxs } from "react/jsx-runtime";
import { Group } from "@visx/group";
import { isEmpty } from "lodash-es";
import { forwardRef } from "react";
import { svgWrapperCSS } from "./GanttChartLineRenderer.sty.js";
import { GanttSegment } from "./GanttSegment.js";
import { ChartStateHandler } from "../../core/components/chart-states/ChartStateHandler.js";
import { getChartStatesTemplates } from "../../core/components/chart-states/utils/get-chart-states-template.js";
import { EmptyStateSlot } from "../../core/slots/chart-states/EmptyStateSlot.js";
import { ErrorStateSlot } from "../../core/slots/chart-states/ErrorStateSlot.js";
import {
  DEFAULT_GANTT_COLORS,
  DISABLED_PATTERN_ANGLE,
  DISABLED_PATTERN_SIZE,
  STRIPES_PATTERN_ID
} from "../constants/defaults.js";
import { useGanttSegmentsContext } from "../context/GanttSegmentsContext.js";
const _GanttChartLineRenderer = forwardRef((props, forwardedRef) => {
  const { width, height: svgHeight, children } = props;
  const { segmentsData } = useGanttSegmentsContext();
  const backgroundSegments = [];
  const { emptyState, errorState } = getChartStatesTemplates(children);
  if (props.showBackground) {
    segmentsData.sort((a, b) => a.start - b.start);
    for (let i = 0; i < segmentsData.length - 1; i++) {
      const currentSegmentEnd = segmentsData[i].end || segmentsData[i].start;
      const nextSegmentStart = segmentsData[i + 1].start;
      if (currentSegmentEnd < nextSegmentStart) {
        backgroundSegments.push({
          start: currentSegmentEnd,
          end: nextSegmentStart
        });
      }
    }
  }
  return /* @__PURE__ */ jsx(
    ChartStateHandler,
    {
      empty: isEmpty(segmentsData),
      height: "100%",
      emptyState,
      errorState,
      isLoading: props.loading === true,
      children: /* @__PURE__ */ jsx("div", { className: svgWrapperCSS, ref: forwardedRef, children: /* @__PURE__ */ jsxs("svg", { width, height: svgHeight, children: [
        backgroundSegments.length > 0 && /* @__PURE__ */ jsx("defs", { children: /* @__PURE__ */ jsxs(
          "pattern",
          {
            id: STRIPES_PATTERN_ID,
            patternUnits: "userSpaceOnUse",
            width: DISABLED_PATTERN_SIZE * 2,
            height: DISABLED_PATTERN_SIZE * 2,
            patternTransform: `rotate(${DISABLED_PATTERN_ANGLE})`,
            children: [
              /* @__PURE__ */ jsx(
                "rect",
                {
                  x: "0",
                  y: "0",
                  width: DISABLED_PATTERN_SIZE,
                  height: DISABLED_PATTERN_SIZE * 2,
                  fill: DEFAULT_GANTT_COLORS.background
                }
              ),
              /* @__PURE__ */ jsx(
                "rect",
                {
                  x: DISABLED_PATTERN_SIZE,
                  y: "0",
                  width: DISABLED_PATTERN_SIZE,
                  height: DISABLED_PATTERN_SIZE * 2,
                  fill: "transparent"
                }
              )
            ]
          }
        ) }),
        backgroundSegments.map((segment, index) => /* @__PURE__ */ jsx(
          GanttSegment,
          {
            containerWidth: width,
            start: segment.start,
            end: segment.end,
            height: svgHeight
          },
          index
        )),
        /* @__PURE__ */ jsx(Group, { children: segmentsData.map((entry, index) => /* @__PURE__ */ jsx(
          GanttSegment,
          {
            containerWidth: width,
            height: svgHeight,
            data: entry,
            start: entry.start,
            end: entry.end
          },
          index
        )) })
      ] }) })
    }
  );
});
const GanttChartLineRenderer = Object.assign(_GanttChartLineRenderer, {
  EmptyState: EmptyStateSlot,
  ErrorState: ErrorStateSlot
});
export {
  GanttChartLineRenderer
};
