// @dynatrace/strato-components/charts/tree-map/constants.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var constants_exports = {};
__export(constants_exports, {
  AGGREGATED_NODE_COLOR: () => AGGREGATED_NODE_COLOR,
  CLUSTER_LABEL_OFFSET_X: () => CLUSTER_LABEL_OFFSET_X,
  CLUSTER_LABEL_OFFSET_Y: () => CLUSTER_LABEL_OFFSET_Y,
  CLUSTER_OTHER_ID: () => CLUSTER_OTHER_ID,
  DARK_LABEL_TEXT: () => DARK_LABEL_TEXT,
  DEFAULT_AGGREGATED_LABEL: () => DEFAULT_AGGREGATED_LABEL,
  DEFAULT_BORDER_COLOR: () => DEFAULT_BORDER_COLOR,
  DEFAULT_CATEGORICAL_COLOR_PALETTE: () => DEFAULT_CATEGORICAL_COLOR_PALETTE,
  DEFAULT_CLUSTER_LABEL_FONTSIZE: () => DEFAULT_CLUSTER_LABEL_FONTSIZE,
  DEFAULT_CLUSTER_LABEL_LINE_HEIGHT: () => DEFAULT_CLUSTER_LABEL_LINE_HEIGHT,
  DEFAULT_FONT_WEIGHT: () => DEFAULT_FONT_WEIGHT,
  DEFAULT_HEIGHT: () => DEFAULT_HEIGHT,
  DEFAULT_LABELS_DISPLAY: () => DEFAULT_LABELS_DISPLAY,
  DEFAULT_LABEL_ALIGN: () => DEFAULT_LABEL_ALIGN,
  DEFAULT_LABEL_BASELINE: () => DEFAULT_LABEL_BASELINE,
  DEFAULT_LABEL_FONTSIZE: () => DEFAULT_LABEL_FONTSIZE,
  DEFAULT_LABEL_LINE_HEIGHT: () => DEFAULT_LABEL_LINE_HEIGHT,
  DEFAULT_LABEL_PADDING: () => DEFAULT_LABEL_PADDING,
  DEFAULT_LOADING_STATE: () => DEFAULT_LOADING_STATE,
  DEFAULT_MIN_NODE_SIZE: () => DEFAULT_MIN_NODE_SIZE,
  DEFAULT_MIN_NODE_SIZE_PARENT: () => DEFAULT_MIN_NODE_SIZE_PARENT,
  DEFAULT_MIN_NODE_SIZE_PARENT_AND_LABELS: () => DEFAULT_MIN_NODE_SIZE_PARENT_AND_LABELS,
  DEFAULT_NAME_ACCESSOR: () => DEFAULT_NAME_ACCESSOR,
  DEFAULT_RANGE_COLOR_PALETTE: () => DEFAULT_RANGE_COLOR_PALETTE,
  DEFAULT_RANGE_LEGEND_COLOR: () => DEFAULT_RANGE_LEGEND_COLOR,
  DEFAULT_ROOT_NODE_NAME: () => DEFAULT_ROOT_NODE_NAME,
  DEFAULT_TOOLBAR_PLACEMENT: () => DEFAULT_TOOLBAR_PLACEMENT,
  DEFAULT_WIDTH: () => DEFAULT_WIDTH,
  DISPLAY_NAME: () => DISPLAY_NAME,
  INCREASED_FONT_WEIGHT: () => INCREASED_FONT_WEIGHT,
  LABELS_DISPLAY: () => LABELS_DISPLAY,
  LABEL_FONT_FAMILY: () => LABEL_FONT_FAMILY,
  LABEL_OPTIONS: () => LABEL_OPTIONS,
  LIGHT_LABEL_TEXT: () => LIGHT_LABEL_TEXT,
  MIN_CLUSTER_HEIGHT: () => MIN_CLUSTER_HEIGHT,
  MIN_CLUSTER_HEIGHT_WITH_VISIBLE_CHILDREN: () => MIN_CLUSTER_HEIGHT_WITH_VISIBLE_CHILDREN,
  MIN_CLUSTER_WIDTH_WITH_VISIBLE_CHILDREN: () => MIN_CLUSTER_WIDTH_WITH_VISIBLE_CHILDREN,
  NAME_ACCESSOR: () => NAME_ACCESSOR,
  NODE_CLUSTER_PADDING: () => NODE_CLUSTER_PADDING,
  NODE_LEAF_PADDING: () => NODE_LEAF_PADDING,
  NODE_WITH_RECT_CLUSTER_PADDING: () => NODE_WITH_RECT_CLUSTER_PADDING,
  NODE_WITH_RECT_OUTER_PADDING: () => NODE_WITH_RECT_OUTER_PADDING,
  NODE_WITH_RECT_PADDING_TOP: () => NODE_WITH_RECT_PADDING_TOP,
  OUT_OF_RANGE_NODE_COLOR: () => OUT_OF_RANGE_NODE_COLOR,
  RECT_BORDER_RADIUS: () => RECT_BORDER_RADIUS,
  RECT_CLUSTER_OPACITY: () => RECT_CLUSTER_OPACITY,
  RECT_STROKE_WIDTH: () => RECT_STROKE_WIDTH,
  ROOT_NODE_WITH_LABELS_OUTER_PADDING: () => ROOT_NODE_WITH_LABELS_OUTER_PADDING,
  TEXT_ALIGN: () => TEXT_ALIGN,
  TEXT_BASELINE: () => TEXT_BASELINE,
  TREEMAP_CHART_ARIA_LABEL: () => TREEMAP_CHART_ARIA_LABEL
});
module.exports = __toCommonJS(constants_exports);
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
var import_constants = require("../core/constants.js");
const DISPLAY_NAME = "TreeMap";
const LABELS_DISPLAY = Object.freeze({
  All: "all",
  Clusters: "clusters",
  ClustersAndValues: "clusters-and-values",
  Nodes: "nodes",
  None: "none"
});
const TEXT_ALIGN = Object.freeze({
  Center: "center",
  End: "end",
  Left: "left",
  Right: "right",
  Start: "start"
});
const TEXT_BASELINE = Object.freeze({
  Alphabetic: "alphabetic",
  Bottom: "bottom",
  Hanging: "hanging",
  Ideographic: "ideographic",
  Middle: "middle",
  Top: "top"
});
const NAME_ACCESSOR = Object.freeze({
  Name: "name",
  Value: "value"
});
const CLUSTER_OTHER_ID = "other-cluster";
const DEFAULT_AGGREGATED_LABEL = "Other";
const DEFAULT_ROOT_NODE_NAME = "root node";
const DEFAULT_WIDTH = "100%";
const DEFAULT_HEIGHT = 300;
const DEFAULT_LOADING_STATE = false;
const DEFAULT_CATEGORICAL_COLOR_PALETTE = "categorical";
const DEFAULT_RANGE_COLOR_PALETTE = "moss";
const DEFAULT_MIN_NODE_SIZE = 250;
const DEFAULT_MIN_NODE_SIZE_PARENT = 1e3;
const DEFAULT_MIN_NODE_SIZE_PARENT_AND_LABELS = 2e3;
const DEFAULT_TOOLBAR_PLACEMENT = "top-right";
const DEFAULT_LABELS_DISPLAY = LABELS_DISPLAY.None;
const DEFAULT_LABEL_ALIGN = TEXT_ALIGN.Center;
const DEFAULT_LABEL_BASELINE = TEXT_BASELINE.Middle;
const DEFAULT_LABEL_FONTSIZE = 12;
const DEFAULT_LABEL_LINE_HEIGHT = 16;
const DEFAULT_CLUSTER_LABEL_FONTSIZE = 12;
const DEFAULT_CLUSTER_LABEL_LINE_HEIGHT = 20;
const CLUSTER_LABEL_OFFSET_Y = DEFAULT_LABEL_FONTSIZE + 2;
const CLUSTER_LABEL_OFFSET_X = 10;
const CLUSTER_LABEL_PADDING = 4;
const CLUSTER_LABEL_PADDING_SUM = CLUSTER_LABEL_PADDING * 2;
const MIN_CLUSTER_HEIGHT = DEFAULT_CLUSTER_LABEL_FONTSIZE + CLUSTER_LABEL_PADDING_SUM;
const DEFAULT_LABEL_PADDING = 4;
const LABEL_FONT_FAMILY = "DT Flow";
const DEFAULT_FONT_WEIGHT = 400;
const INCREASED_FONT_WEIGHT = 500;
const DEFAULT_NAME_ACCESSOR = void 0;
const DARK_LABEL_TEXT = "#000000";
const LIGHT_LABEL_TEXT = "#ffffff";
const RECT_BORDER_RADIUS = 2;
const RECT_STROKE_WIDTH = 1;
const RECT_CLUSTER_OPACITY = 0.15;
const NODE_LEAF_PADDING = 1;
const NODE_CLUSTER_PADDING = 2;
const NODE_WITH_RECT_CLUSTER_PADDING = 7;
const NODE_WITH_RECT_PADDING_TOP = DEFAULT_CLUSTER_LABEL_FONTSIZE + CLUSTER_LABEL_PADDING_SUM;
const ROOT_NODE_WITH_LABELS_OUTER_PADDING = 3;
const NODE_WITH_RECT_OUTER_PADDING = 3;
const MIN_VISIBLE_NODE_HEIGHT_AND_WIDTH = 1;
const ADDITIONAL_CLUSTER_HEIGHT_SPACING = 2;
const ADDITIONAL_CLUSTER_WIDTH_SPACING = 4;
const MIN_CLUSTER_HEIGHT_WITH_VISIBLE_CHILDREN = DEFAULT_CLUSTER_LABEL_FONTSIZE + MIN_VISIBLE_NODE_HEIGHT_AND_WIDTH + 2 * RECT_STROKE_WIDTH + CLUSTER_LABEL_PADDING_SUM + ADDITIONAL_CLUSTER_HEIGHT_SPACING;
const MIN_CLUSTER_WIDTH_WITH_VISIBLE_CHILDREN = MIN_VISIBLE_NODE_HEIGHT_AND_WIDTH + 2 * RECT_STROKE_WIDTH + ADDITIONAL_CLUSTER_WIDTH_SPACING;
const AGGREGATED_NODE_COLOR = import_colors.default.Background.Container.Neutral.Accent;
const OUT_OF_RANGE_NODE_COLOR = import_colors.default.Background.Container.Neutral.Accent;
const DEFAULT_RANGE_LEGEND_COLOR = import_constants.SCALE_FALLBACK_COLOR;
const DEFAULT_BORDER_COLOR = import_colors.default.Border.Neutral.Accent;
const LABEL_OPTIONS = {
  Ellipsis: "ellipsis",
  EllipsisWithValue: "ellipsis-with-value",
  Fallback: "fallback",
  Inline: "inline",
  Multi: "multi",
  Name: "name"
};
const TREEMAP_CHART_ARIA_LABEL = "Tree Map Chart";
