// @dynatrace/strato-components/charts/tree-map/defaults.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var defaults_exports = {};
__export(defaults_exports, {
  DEFAULT_LEGEND: () => DEFAULT_LEGEND,
  defaultTreeMapConfig: () => defaultTreeMapConfig
});
module.exports = __toCommonJS(defaults_exports);
var import_constants = require("./constants.js");
var import_constants2 = require("../core/constants.js");
const DEFAULT_LEGEND = {
  position: import_constants2.LEGEND_POSITION.Auto,
  hidden: false,
  ratio: import_constants2.defaultRatio
};
const defaultTreeMapConfig = {
  colorPalette: import_constants.DEFAULT_CATEGORICAL_COLOR_PALETTE,
  formatter: (v) => v.toString(),
  truncationMode: import_constants2.DEFAULT_TRUNCATION_MODE,
  loading: import_constants.DEFAULT_LOADING_STATE,
  legend: DEFAULT_LEGEND,
  labelsDisplay: import_constants.DEFAULT_LABELS_DISPLAY,
  nameAccessor: import_constants.DEFAULT_NAME_ACCESSOR,
  min: void 0,
  max: void 0
};
