// @dynatrace/strato-components/charts/tree-map/TreeMap.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TreeMap_exports = {};
__export(TreeMap_exports, {
  TreeMap: () => TreeMap
});
module.exports = __toCommonJS(TreeMap_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_TreeMapChartLayout = require("./components/TreeMapChartLayout.js");
var import_constants = require("./constants.js");
var import_useTreeMapChartConfig = require("./hooks/useTreeMapChartConfig.js");
var import_TreeMapChartProviders = require("./providers/TreeMapChartProviders.js");
var import_DownloadCSV = require("./slots/DownloadCSV.js");
var import_Toolbar = require("./slots/Toolbar.js");
var import_Tooltip = require("./slots/Tooltip.js");
var import_TreeMapChartLegend = require("./slots/TreeMapChartLegend.js");
var import_build_and_download_csv = require("./utils/build-and-download-csv.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_Intent = require("../../core/slots/Intent/Intent.js");
var import_defaultChartValues_css = require("../../core/styles/defaultChartValues.sty.js");
var import_useAutoLegendRefresh = require("../core/components/chart-layout/hooks/useAutoLegendRefresh.js");
var import_get_chart_states_template = require("../core/components/chart-states/utils/get-chart-states-template.js");
var import_EmptyStateSlot = require("../core/slots/chart-states/EmptyStateSlot.js");
var import_ErrorStateSlot = require("../core/slots/chart-states/ErrorStateSlot.js");
var import_ColorRule = require("../core/slots/ColorRule.js");
var import_chart_types = require("../core/types/chart-types.js");
const COMPONENT_NAME = "TreeMap";
const TreeMapRoot = (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    data,
    children,
    seriesActions,
    width = import_constants.DEFAULT_WIDTH,
    height = import_constants.DEFAULT_HEIGHT,
    className,
    style,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "data-testid": dataTestId = "treemap-chart-root",
    unit
  } = props;
  const [behavioralTrackingProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(props);
  const config = (0, import_useTreeMapChartConfig.useTreeMapChartConfig)(props);
  const { loading } = config;
  const emptyData = (0, import_lodash_es.isEmpty)(data) || data.tree.nodes === void 0 || data.tree.nodes.length === 0;
  const { emptyState, errorState } = (0, import_get_chart_states_template.getChartStatesTemplates)(children);
  const treeMapRef = (0, import_react.useRef)(null);
  const chartLayoutRef = (0, import_useAutoLegendRefresh.useAutoLegendRefresh)(data);
  (0, import_react.useImperativeHandle)(forwardedRef, () => {
    let exposedRef = {};
    if (treeMapRef.current) {
      exposedRef = {
        element: treeMapRef.current,
        getConfig: () => JSON.stringify({ type: import_chart_types.ChartType.Treemap, ...config }),
        downloadData: () => (0, import_build_and_download_csv.buildAndDownloadCsv)(data.tree)
      };
    }
    return exposedRef;
  });
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_TreeMapChartProviders.TreeMapChartProviders,
    {
      data,
      seriesActions,
      unit,
      ...config,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_TreeMapChartLayout.TreeMapChartLayout,
        {
          ...behavioralTrackingProps,
          chartLayoutRef,
          ref: treeMapRef,
          width,
          height,
          emptyState,
          errorState,
          isEmpty: emptyData,
          loading,
          className: (0, import_clsx.default)(import_defaultChartValues_css.defaultChartSize, className),
          style,
          "data-dtrum-mask": dataDtrumMask,
          "data-dtrum-allow": dataDtrumAllow,
          "data-testid": dataTestId,
          "data-dt-component": COMPONENT_NAME
        }
      )
    }
  );
});
const TreeMap = Object.assign(TreeMapRoot, {
  Legend: import_TreeMapChartLegend.Legend,
  Tooltip: import_Tooltip.Tooltip,
  Intent: import_Intent.Intent,
  Toolbar: import_Toolbar.Toolbar,
  DownloadCSV: import_DownloadCSV.TreeMapDownloadCSVSlot,
  EmptyState: import_EmptyStateSlot.EmptyStateSlot,
  ErrorState: import_ErrorStateSlot.ErrorStateSlot,
  ColorRule: import_ColorRule.ColorRule
});
