// @dynatrace/strato-components/esm/charts/core/components/axes/components/PanningControls/PanningControls.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { useIntl } from "react-intl";
import {
  ArrowSmallLeftIcon,
  ArrowSmallRightIcon
} from "@dynatrace/strato-icons";
import { PANNING_CONTROLS_MESSAGES } from "./messages.js";
import { panningArrow, panningButtons } from "./PanningControls.sty.js";
import { Tooltip } from "../../../../../../overlays/tooltip/Tooltip.js";
const PANNING_STEP = 40;
const PanningControls = (props) => {
  const { showRightControl, showLeftControl, onClickPan } = props;
  const intl = useIntl();
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    showLeftControl && /* @__PURE__ */ jsx(
      "button",
      {
        "aria-label": intl.formatMessage(
          PANNING_CONTROLS_MESSAGES.buttonLeftLabel
        ),
        className: panningButtons({ direction: "toLeft" }),
        onClick: () => onClickPan(PANNING_STEP),
        children: /* @__PURE__ */ jsx(
          Tooltip,
          {
            text: intl.formatMessage(PANNING_CONTROLS_MESSAGES.panLeft),
            placement: "bottom",
            children: /* @__PURE__ */ jsx(ArrowSmallLeftIcon, { size: "default", className: panningArrow })
          }
        )
      }
    ),
    showRightControl && /* @__PURE__ */ jsx(
      "button",
      {
        "aria-label": intl.formatMessage(
          PANNING_CONTROLS_MESSAGES.buttonRightLabel
        ),
        className: panningButtons({ direction: "toRight" }),
        onClick: () => onClickPan(-PANNING_STEP),
        children: /* @__PURE__ */ jsx(
          Tooltip,
          {
            text: intl.formatMessage(PANNING_CONTROLS_MESSAGES.panRight),
            placement: "bottom",
            children: /* @__PURE__ */ jsx(ArrowSmallRightIcon, { size: "default", className: panningArrow })
          }
        )
      }
    )
  ] });
};
export {
  PanningControls
};
