// @dynatrace/strato-components/esm/forms/calendar/utils/helpers.js
import { TZDateMini } from "@date-fns/tz";
import {
  addDays,
  addWeeks,
  compareAsc,
  endOfDay,
  endOfWeek,
  intervalToDuration,
  isBefore,
  isEqual,
  isSameDay,
  isWithinInterval,
  lastDayOfMonth,
  lastDayOfWeek,
  startOfDay,
  startOfMonth,
  startOfWeek
} from "date-fns";
import { getWeekStartByLocale } from "weekstart";
import {
  getTimezone,
  getRegionalFormat
} from "@dynatrace-sdk/user-preferences";
import { isTimeframe, isTimeValue } from "../../../core/time/helpers.js";
import {
  getTimeValueFromSimpleValue,
  sortTimeframeValues,
  eachDayOfInterval,
  timeValueToDate
} from "../../forms-core/time/helpers.js";
import {
  isExpressionString,
  mapToPartialTimeframe,
  removePrecision,
  removePrecisionFromTimeValue
} from "../../forms-core/time/timeParser.js";
import {
  NUMBER_OF_WEEKDAYS,
  NUMBER_OF_WEEK_ROWS
} from "../types/calendarConfiguration.js";
function getDays(activeDate, weekStartsOn) {
  const monthStart = startOfMonth(activeDate);
  const firstWeekStart = startOfWeek(monthStart, {
    weekStartsOn
  });
  const monthEnd = lastDayOfMonth(monthStart);
  const lastWeekEnd = lastDayOfWeek(monthEnd, { weekStartsOn });
  let prevMonthDays = getSiblingMonthDays(monthStart, firstWeekStart, -1);
  const currentMonthDays = toDayArray(
    eachDayOfInterval({
      start: monthStart,
      end: monthEnd
    }),
    true
  );
  const nextMonthDays = getSiblingMonthDays(monthEnd, lastWeekEnd, 1);
  const days = [...prevMonthDays, ...currentMonthDays, ...nextMonthDays];
  while (days.length < NUMBER_OF_WEEKDAYS * NUMBER_OF_WEEK_ROWS) {
    if (!prevMonthDays.length) {
      prevMonthDays = toDayArray(
        eachDayOfInterval({
          // Typecast to TZDate since TypeScript doesn't infer addWeeks return type directly
          start: addWeeks(firstWeekStart, -1),
          end: addDays(firstWeekStart, -1)
        })
      );
      days.unshift(...prevMonthDays);
    } else {
      days.push(
        ...toDayArray(
          eachDayOfInterval({
            // Typecast to TZDate since TypeScript doesn't infer addWeeks return type directly
            start: addDays(lastWeekEnd, 1),
            end: addWeeks(lastWeekEnd, 1)
          })
        )
      );
    }
  }
  return days;
}
function getSiblingMonthDays(monthStartEnd, weekStartEnd, direction) {
  if (isEqual(monthStartEnd, weekStartEnd)) {
    return [];
  }
  const siblingMonthStartEnd = addDays(monthStartEnd, direction);
  const sortedDays = [siblingMonthStartEnd, weekStartEnd].sort(
    (a, b) => compareAsc(a, b)
  );
  return toDayArray(
    eachDayOfInterval({
      start: sortedDays[0],
      end: sortedDays[1]
    })
  );
}
function toDayArray(array, isCurrentMonthDay = false) {
  return array.map(
    (date) => ({
      date,
      isCurrentMonthDay
    })
  );
}
function convertToCalendarStructure(type = "single", value, min, max, precision) {
  if (type === "single" && !value) {
    return null;
  }
  if (typeof value === "string" || value instanceof Date) {
    const stringDate = typeof value === "string" ? value : value.toISOString();
    const timeOffset = getTimeOffsetFromInitialValue(
      stringDate,
      "single",
      min,
      max,
      precision
    );
    return getTimeValueFromSimpleValue(
      value,
      timeOffset,
      void 0,
      void 0,
      min,
      max,
      precision
    ) ?? null;
  }
  if (isTimeValue(value)) {
    return removePrecisionFromTimeValue(value, precision);
  }
  if (value) {
    const timeframe = mapToPartialTimeframe(value);
    const timeOffset = getTimeOffsetFromInitialValue(
      timeframe,
      "range",
      min,
      max,
      precision
    );
    const from = getTimeValueFromSimpleValue(
      timeframe?.from?.value,
      timeOffset,
      "from",
      timeframe?.from?.type === "expression",
      min,
      void 0,
      precision
    );
    const to = getTimeValueFromSimpleValue(
      timeframe?.to?.value,
      timeOffset,
      "to",
      timeframe?.to?.type === "expression",
      void 0,
      max,
      precision
    );
    return sortTimeframeValues({ from, to }, min, max, precision);
  }
  return { from: null, to: null };
}
function getRangePosition(newDate, from, to) {
  if (from && isSameDay(newDate, from)) {
    return "first";
  }
  if (to && isSameDay(newDate, to)) {
    return "last";
  }
  if (isWithinInterval(newDate, {
    start: startOfDay(from),
    end: endOfDay(to)
  })) {
    return "inbetween";
  }
}
function isRelativeDate(highlightType, selectedDate) {
  switch (highlightType) {
    case "first": {
      if (selectedDate.from) {
        return selectedDate.from.type === "iso8601" ? "absolute" : "relative";
      }
      return void 0;
    }
    case "last": {
      if (selectedDate.to) {
        return selectedDate.to.type === "iso8601" ? "absolute" : "relative";
      }
      if (!selectedDate.to && selectedDate.from) {
        return selectedDate.from.type === "iso8601" ? "absolute" : "relative";
      }
      return void 0;
    }
    case "same": {
      if (selectedDate.from?.type === "expression" && selectedDate.to?.type === "expression") {
        return "relative";
      }
      if (selectedDate.from?.type === "expression" && selectedDate.to?.type === "iso8601") {
        return "relative-absolute";
      }
      if (selectedDate.from?.type === "iso8601" && selectedDate.to?.type === "expression") {
        return "absolute-relative";
      }
      return "absolute";
    }
    case "inbetween":
    default: {
      return void 0;
    }
  }
}
function getHighlightType(from, to, newDate) {
  if (to && from && isSameDay(from, newDate) && isSameDay(to, newDate) || !to && from && isSameDay(from, newDate) || !from && to && isSameDay(to, newDate)) {
    return "same";
  }
  if (from && to && isBefore(from, to)) {
    return getRangePosition(newDate, from, to);
  }
  if (from && to && isBefore(to, from)) {
    return getRangePosition(newDate, to, from);
  }
  return void 0;
}
const getHighlightPropsFromTimeframe = (date, selectedDate, hoveredDate) => {
  const { from: selectedFrom, to: selectedTo } = selectedDate;
  if (!selectedFrom && !selectedTo) {
    return {};
  }
  const relativeDate = TZDateMini.tz(getTimezone()).getTime();
  const from = timeValueToDate(selectedFrom, relativeDate);
  const to = timeValueToDate(selectedTo, relativeDate);
  const isSelected = !!(from && isSameDay(date, from) || to && isSameDay(date, to));
  const safeFromDate = from ?? hoveredDate;
  const safeToDate = to ?? hoveredDate;
  const highlightType = getHighlightType(safeFromDate, safeToDate, date);
  const isEntireRangeSelected = !!selectedFrom && !!selectedTo && !!highlightType;
  const highlightColor = isSelected || isEntireRangeSelected ? "primary" : "neutral";
  const isRelative = isSelected ? isRelativeDate(highlightType, selectedDate) : void 0;
  return {
    highlightType,
    highlightColor,
    isSelected,
    isIncompleteSelection: !isEntireRangeSelected,
    selectionVariant: isRelative
  };
};
const getHighlightingProps = (date, selectedDate, type, hoveredDate = null) => {
  if (!selectedDate) {
    return {};
  }
  if (isTimeframe(selectedDate, type)) {
    return getHighlightPropsFromTimeframe(date, selectedDate, hoveredDate);
  }
  if (isTimeValue(selectedDate)) {
    const newSelectedDate = timeValueToDate(selectedDate, Date.now());
    if (!newSelectedDate) {
      return {};
    }
    const isSelected = isSameDay(date, newSelectedDate);
    if (isSelected) {
      return {
        isSelected: true,
        highlightType: "same",
        highlightColor: "primary",
        isIncompleteSelection: false,
        selectionVariant: selectedDate.type === "expression" ? "relative" : "absolute"
      };
    }
    return {
      isSelected: false
    };
  }
  return {};
};
function getWeekStartOrEnd(indicator, activeDate) {
  const region = getRegionalFormat();
  const weekStartsOn = getWeekStartByLocale(region);
  const start = indicator === "startOfWeek" ? startOfWeek(activeDate, { weekStartsOn }) : endOfWeek(activeDate, { weekStartsOn });
  return (intervalToDuration({
    start,
    end: activeDate
  }).days ?? 0) * -1;
}
function getTimeOffsetFromInitialValue(value, type = "single", min = null, max = null, precision = "milliseconds") {
  if (isTimeframe(value, type)) {
    if (!value) {
      return { from: "startOfDay", to: "endOfDay" };
    }
    let from = value["from"]?.absoluteDate;
    let to = value["to"]?.absoluteDate;
    const configuredTimezone = getTimezone();
    const fromIsAtStartOfDay = from && from === startOfDay(new TZDateMini(from, configuredTimezone)).toISOString();
    const fromIsMin = min && from && removePrecision(from, precision) === removePrecision(min?.absoluteDate, precision);
    if (!from || fromIsAtStartOfDay || value["from"]?.type === "expression" || fromIsMin) {
      from = "startOfDay";
    }
    const toIsAtEndOfDay = to && to === endOfDay(new TZDateMini(to, configuredTimezone)).toISOString();
    const toIsMax = max && to && removePrecision(to, precision) === removePrecision(max?.absoluteDate, precision);
    if (!to || toIsAtEndOfDay || value.to?.type === "expression" || toIsMax) {
      to = "endOfDay";
    }
    return {
      from,
      to
    };
  }
  if (precision === "day") {
    return "startOfDay";
  }
  const offset = isTimeValue(value) ? value.value : value;
  if (typeof offset === "string") {
    return isExpressionString(offset) ? "startOfDay" : offset;
  }
  return "startOfDay";
}
export {
  convertToCalendarStructure,
  getDays,
  getHighlightingProps,
  getTimeOffsetFromInitialValue,
  getWeekStartOrEnd
};
