// @dynatrace/strato-components/esm/charts/core/components/interactions-overlay/interactions/pan/dragInteractions.js
import { isUndefined } from "lodash-es";
import { centerPositionVertically } from "../../utils/center-position-vertically.js";
const panDragStart = ({
  panDeltaRef,
  onPanStart
}) => {
  onPanStart?.({ triggeredBy: "pointer" });
  panDeltaRef.current = 0;
};
const panDragMove = ({
  event,
  height,
  panDeltaRef,
  onPanChange
}) => {
  const {
    startHoldingAt: startHoldingAtProp,
    currentPosition: currentPositionProp
  } = event;
  const startHoldingAt = centerPositionVertically(startHoldingAtProp, height);
  const currentPosition = centerPositionVertically(currentPositionProp, height);
  if (!isUndefined(panDeltaRef.current)) {
    const fullInteractionDelta = currentPosition.relative.x - startHoldingAt.relative.x;
    const currentDelta = fullInteractionDelta - panDeltaRef.current;
    panDeltaRef.current = fullInteractionDelta;
    onPanChange?.({ triggeredBy: "pointer", delta: -currentDelta });
  }
};
const panDragEnd = ({
  panDeltaRef,
  onPanEnd
}) => {
  panDeltaRef.current = void 0;
  onPanEnd?.({ triggeredBy: "pointer" });
};
const panDragStartWithMiddleButton = ({
  panDeltaRef,
  onPanStart
}) => {
  onPanStart?.({ triggeredBy: "pointer" });
  panDeltaRef.current = 0;
};
const panDragMoveWithMiddleButton = ({
  event,
  panDeltaRef,
  onPanChange
}) => {
  if (!isUndefined(panDeltaRef.current)) {
    const { currentPosition, startHoldingAt } = event;
    const fullInteractionDelta = currentPosition.relative.x - startHoldingAt.relative.x;
    const currentDelta = fullInteractionDelta - panDeltaRef.current;
    panDeltaRef.current = fullInteractionDelta;
    onPanChange?.({ triggeredBy: "pointer", delta: -currentDelta });
  }
};
const panDragEndWithMiddleButton = ({
  event,
  panDeltaRef,
  lastMousePositionRef,
  onPanEnd
}) => {
  panDeltaRef.current = void 0;
  lastMousePositionRef.current = event.releasedAt;
  onPanEnd?.({ triggeredBy: "pointer" });
};
export {
  panDragEnd,
  panDragEndWithMiddleButton,
  panDragMove,
  panDragMoveWithMiddleButton,
  panDragStart,
  panDragStartWithMiddleButton
};
