// @dynatrace/strato-components/esm/charts/core/components/interactions-overlay/interactions/pan/keyboardInteractions.js
import { isUndefined } from "lodash-es";
import { LARGE_STEP_FACTOR, SMALL_STEP_FACTOR } from "../../constants.js";
const panKeyboardEscape = ({
  selectionArea,
  eventCancelled,
  onSelectEnd,
  lastMousePositionRef,
  pinned,
  onHover,
  onTogglePin,
  onModeChange,
  onPanEnd
}) => {
  onPanEnd?.({ triggeredBy: "keyboard" });
  if (!isUndefined(selectionArea)) {
    eventCancelled.current = true;
    onSelectEnd?.({ cancelled: true });
    if (lastMousePositionRef.current) {
      onHover?.({
        triggeredBy: "pointer",
        position: lastMousePositionRef.current
      });
    }
  }
  if (pinned) {
    onTogglePin?.({
      triggeredBy: "keyboard",
      pin: false,
      position: lastMousePositionRef.current
    });
  }
  onModeChange?.({ mode: "explore", keyboard: false });
};
const panKeyboardLeft = ({
  shiftKey,
  width,
  onPanChange
}) => {
  const factor = shiftKey ? -LARGE_STEP_FACTOR : -SMALL_STEP_FACTOR;
  const delta = width * factor;
  onPanChange?.({ triggeredBy: "keyboard", delta });
};
const panKeyboardRight = ({
  shiftKey,
  width,
  onPanChange
}) => {
  const factor = shiftKey ? LARGE_STEP_FACTOR : SMALL_STEP_FACTOR;
  const delta = width * factor;
  onPanChange?.({ triggeredBy: "keyboard", delta });
};
export {
  panKeyboardEscape,
  panKeyboardLeft,
  panKeyboardRight
};
