// @dynatrace/strato-components/esm/filters/filter-field/utils/conversion/lezer-tree-conversion/parent-node/group.js
import { convertFilterStatement } from "./statement.js";
import { isDev } from "../../../../../../core/utils/is-node-environment.js";
import { postSpotErrorMark } from "../../../../decorations/decoration-marks.js";
import { getColoredSearchOperator } from "../../../validation-utils/validation-helper-functions.js";
import { convertLogicalOperator } from "../logical-operator.js";
import { convertError } from "../value/error.js";
function convertFilterGroup(node, doc, numberParser, internalValidatorMap, customTypes, parserConfig, context) {
  const { from, to } = node;
  const cursor = node.node.cursor();
  const convertedNode = {
    type: "Group",
    children: [],
    range: {
      from,
      to
    },
    explicit: false,
    explicitLogicalOperator: false,
    logicalOperator: "AND"
  };
  const errors = [];
  let hasOpeningBracket = false;
  let hasClosingBracket = false;
  let hasChildNodes = false;
  cursor.iterate((childNode) => {
    const childName = childNode.name;
    switch (childName) {
      case "Filter":
        break;
      case "FilterGroup": {
        if (node.name !== "Filter" && childNode.from === from && childNode.to === to) {
          break;
        }
        hasChildNodes = true;
        const { convertedNode: groupNode, errors: groupErrors } = convertFilterGroup(
          childNode,
          doc,
          numberParser,
          internalValidatorMap,
          customTypes,
          parserConfig,
          "FilterGroup"
        );
        convertedNode.children.push(groupNode);
        errors.push(...groupErrors);
        return false;
      }
      case "FilterStatement": {
        hasChildNodes = true;
        const { convertedNode: statementNode, errors: statementErrors } = convertFilterStatement(
          childNode,
          doc,
          numberParser,
          internalValidatorMap,
          customTypes,
          parserConfig,
          context
        );
        convertedNode.children.push(statementNode);
        errors.push(...statementErrors);
        return false;
      }
      case "LogicalOperator": {
        hasChildNodes = true;
        const { convertedNode: operatorNode, errors: logicalOperatorErrors } = convertLogicalOperator(childNode, doc);
        if (logicalOperatorErrors.length > 0) {
          convertedNode.children.push({
            type: "Statement",
            range: operatorNode.range,
            key: operatorNode
          });
        } else {
          convertedNode.children.push(operatorNode);
        }
        errors.push(...logicalOperatorErrors);
        return false;
      }
      case "OpeningBracket":
        if (node.name !== "Filter") {
          hasOpeningBracket = true;
          convertedNode.explicit = true;
        }
        break;
      case "ClosingBracket":
        if (node.name !== "Filter") {
          hasClosingBracket = true;
          convertedNode.explicit = true;
        }
        break;
      case "\u26A0":
        if (node.name === "Filter") {
          if (childNode.node.firstChild) {
            const { convertedNode: errorNode } = convertError(
              cursor,
              childNode,
              doc
            );
            convertedNode.children.push(errorNode);
            errors.push({
              range: errorNode.range,
              message: "incompleteGroupError"
            });
          } else {
            const { convertedNode: errorNode } = convertError(
              cursor,
              childNode,
              doc
            );
            convertedNode.children.push(errorNode);
            if (parserConfig?.searchConversion) {
              errors.push({
                range: errorNode.range,
                message: {
                  type: "incompleteStatementWithEnabledSearchError",
                  details: getColoredSearchOperator()
                }
              });
            } else {
              errors.push({
                range: errorNode.range,
                message: "incompleteStatementError"
              });
            }
          }
        }
        return false;
      default:
        if (isDev()) {
          console.log(
            `${doc.toString()} | Syntax node '${childName}' cannot be converted. Please make sure to handle all nodes accordingly.`
          );
        }
    }
  });
  const groupError = {
    type: "Error",
    range: { from, to },
    value: "",
    textValue: ""
  };
  const emptyBracketGroup = hasOpeningBracket && hasClosingBracket && !hasChildNodes;
  if (emptyBracketGroup) {
    groupError.range = { from: from + 1, to: to - 1 };
    convertedNode.children.push(groupError);
    errors.push({
      range: { from, to },
      message: "emptyGroupError"
    });
  } else if (hasOpeningBracket && !hasClosingBracket) {
    groupError.range = { from: to, to };
    convertedNode.children.push(groupError);
    errors.push({
      range: { from: to - 1, to },
      message: "incompleteGroupError",
      decoration: postSpotErrorMark
    });
  }
  return { convertedNode, errors };
}
export {
  convertFilterGroup
};
