// @dynatrace/strato-components/esm/filters/filter-field/utils/conversion/lezer-tree-conversion/parent-node/inclusion.js
import { isDev } from "../../../../../../core/utils/is-node-environment.js";
import { getTextFromRange } from "../../../get-text.js";
import { convertToErrorNode, updateNodeValues } from "../node-mutations.js";
import { convertBoolean } from "../value/boolean.js";
import { convertDuration } from "../value/duration.js";
import { convertIdentifier } from "../value/identifier.js";
import { convertNumber } from "../value/number.js";
import { convertSimpleValue } from "../value/simple-value.js";
function convertInclusionValue(node, doc, numberParser, customTypes, parserConfig) {
  const { from, to } = node;
  const cursor = node.node.cursor();
  const convertedNode = {
    type: "List",
    range: {
      from,
      to
    },
    textValue: getTextFromRange(node, doc),
    value: []
  };
  const errors = [];
  const parentheses = { opening: false, closing: false };
  let highlightFullListAsError = false;
  cursor.iterate((childNode) => {
    const nodeName = childNode.name;
    const { from: childFrom, to: childTo } = childNode;
    switch (nodeName) {
      case "FilterIncludesValue":
      case "List": {
        break;
      }
      case "OpeningBracket":
        parentheses.opening = true;
        break;
      case "ClosingBracket":
        parentheses.closing = true;
        break;
      case "ListValue": {
        const {
          convertedNode: valueNode,
          errors: valueErrors,
          positionOperators
        } = convertSimpleValue(childNode, doc, numberParser, customTypes, parserConfig);
        if (positionOperators) {
          convertToErrorNode(valueNode);
          for (const range of positionOperators.ranges) {
            errors.push({
              range,
              message: "invalidWildCardOperatorInList"
            });
          }
        }
        convertedNode.value.push(valueNode);
        errors.push(...valueErrors);
        return false;
      }
      case "Identifier":
      case "String":
      case "IncompleteString": {
        const {
          convertedNode: valueNode,
          positionOperators: { endsWithOperator, startsWithOperator }
        } = convertIdentifier(childNode, doc, nodeName);
        if (endsWithOperator) {
          convertToErrorNode(valueNode);
          errors.push({
            range: endsWithOperator,
            message: "unsupportedPositionOperator"
          });
        }
        if (startsWithOperator) {
          convertToErrorNode(valueNode);
          errors.push({
            range: startsWithOperator,
            message: "unsupportedPositionOperator"
          });
        }
        convertedNode.value.push(valueNode);
        break;
      }
      case "Number": {
        const { convertedNode: numberNode } = convertNumber(
          childNode,
          doc,
          numberParser
        );
        convertedNode.value.push(numberNode);
        break;
      }
      case "Boolean": {
        const { convertedNode: booleanNode } = convertBoolean(childNode, doc);
        convertedNode.value.push(booleanNode);
        break;
      }
      case "Duration": {
        const { convertedNode: durationNode } = convertDuration(childNode, doc);
        convertedNode.value.push(durationNode);
        break;
      }
      case "PositionOperator":
      case "InvalidPositionOperator": {
        const text = getTextFromRange(childNode, doc);
        const errorNode = {
          type: "Error",
          range: { from: childFrom, to: childTo },
          textValue: text,
          value: text
        };
        convertedNode.value.push(errorNode);
        errors.push({
          range: { from: childFrom, to: childTo },
          message: "unsupportedPositionOperator"
        });
        break;
      }
      case "InvalidInclusionValue":
        highlightFullListAsError = true;
        convertToErrorNode(convertedNode);
        updateNodeValues(convertedNode, getTextFromRange(childNode, doc));
        break;
      case "\u26A0":
        highlightFullListAsError = true;
        break;
      default:
        if (isDev()) {
          console.log(
            `${doc.toString()} | Syntax node '${childNode.name}' is not yet handled in the list conversion. Please make sure to handle all nodes accordingly.`
          );
        }
    }
  });
  if (convertedNode.value.length === 0) {
    convertedNode.textValue = getTextFromRange(convertedNode.range, doc);
    convertToErrorNode(convertedNode);
    errors.push({
      range: { from, to },
      message: "emptyInclusionValue"
    });
  } else if (highlightFullListAsError) {
    convertedNode.textValue = getTextFromRange(convertedNode.range, doc);
    convertToErrorNode(convertedNode);
    errors.push({
      range: { from, to },
      message: "invalidInclusionValue"
    });
  }
  return {
    convertedNode,
    errors
  };
}
export {
  convertInclusionValue
};
