// @dynatrace/strato-components/esm/forms/forms-core/components/field-set/FieldSet.js
import { jsx, jsxs } from "react/jsx-runtime";
import classNames from "clsx";
import {
  forwardRef,
  useMemo,
  useReducer,
  Children
} from "react";
import { isElement } from "react-is";
import * as styles from "./FieldSet.sty.js";
import { useBehavioralTrackingProps } from "../../../../core/hooks/useBehavioralTrackingProps.js";
import { mergeProps } from "../../../../core/utils/merge-props.js";
import { parseBoolean as _parseBoolean } from "../../../../core/utils/parse-boolean.js";
import { Flex } from "../../../../layouts/flex/Flex.js";
import { _FormFieldContext as FormFieldContext } from "../../contexts/FormFieldContext.js";
import { _formFieldReducer } from "../../state/formfield-state-reducer.js";
const COMPONENT_NAME = "FieldSet";
const RootFieldSet = forwardRef(
  (props, forwardedRef) => {
    const {
      children,
      disabled: propDisabled,
      name,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "aria-disabled": ariaDisabledProp,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
    const [formFieldState] = useReducer(_formFieldReducer, {});
    const slots = useMemo(() => getSlots(children), [children]);
    return /* @__PURE__ */ jsx(
      FormFieldContext.Provider,
      {
        value: {
          disabled: formFieldState.disabled ?? propDisabled,
          ariaDisabled: formFieldState.ariaDisabled ?? _parseBoolean(ariaDisabledProp),
          required: formFieldState.required,
          dispatch: () => void 0
        },
        children: /* @__PURE__ */ jsx(
          "div",
          {
            "data-testid": dataTestId,
            "data-dtrum-mask": dataDtrumMask,
            "data-dtrum-allow": dataDtrumAllow,
            className: consumerClassName,
            style: consumerStyle,
            ref: forwardedRef,
            ...mergeProps(restProps, behavioralTrackingProps),
            "data-dt-component": COMPONENT_NAME,
            children: /* @__PURE__ */ jsxs("fieldset", { name, className: styles.fieldset, children: [
              slots.legend,
              /* @__PURE__ */ jsx(Flex, { gap: 12, flexDirection: "column", children: slots.rest })
            ] })
          }
        )
      }
    );
  }
);
RootFieldSet.displayName = COMPONENT_NAME;
function getSlots(children) {
  let legend;
  const rest = [];
  for (const child of Children.toArray(children)) {
    if (isElement(child) && child.type === Legend) {
      legend = child;
    } else {
      rest.push(child);
    }
  }
  return { legend, rest };
}
const Legend = forwardRef(
  (props, ref) => {
    const { style, className, ...rest } = props;
    const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(rest);
    return /* @__PURE__ */ jsx(
      "legend",
      {
        className: classNames(styles.legend, className),
        style,
        ref,
        ...mergeProps(restProps, behavioralTrackingProps),
        children: props.children
      }
    );
  }
);
Legend.displayName = "FieldSet.Legend";
const FieldSet = Object.assign(RootFieldSet, {
  Legend
});
export {
  FieldSet
};
