// @dynatrace/strato-components/charts/core/components/tooltip/InteractiveChartTooltip/components/BasicChartTooltip/BasicChartTooltip.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var BasicChartTooltip_exports = {};
__export(BasicChartTooltip_exports, {
  BasicChartTooltip: () => BasicChartTooltip
});
module.exports = __toCommonJS(BasicChartTooltip_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_dom = require("react-dom");
var import_BasicChartTooltip_css = require("./BasicChartTooltip.sty.js");
var import_useOverlayStacking = require("../../../../../../../core/hooks/useOverlayStacking.js");
var import_OverlayStackingProvider = require("../../../../../../../core/providers/OverlayStackingProvider.js");
var import_DataDtrumContext = require("../../../../../context/DataDtrumContext.js");
const BasicChartTooltip = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    placement = "left",
    children,
    chartRef,
    onMouseLeave,
    pinned = false
  } = props;
  const { dataDtrumMask, dataDtrumAllow } = (0, import_DataDtrumContext.useDataDtrumContext)();
  const overlayStacking = (0, import_useOverlayStacking.useOverlayStacking)(0, "overlay");
  const chartTooltipContent = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      className: (0, import_BasicChartTooltip_css.wrapper)({ isPinned: pinned }),
      style: { zIndex: overlayStacking.style.zIndex },
      "data-testid": "chart-tooltip",
      role: "tooltip",
      ref: forwardedRef,
      "data-placement": placement,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      onMouseLeave,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayStackingProvider.OverlayStackingProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: import_BasicChartTooltip_css.chartTooltip, children }) })
    }
  );
  const placementNode = chartRef ? chartRef.current : document.body;
  return (0, import_react_dom.createPortal)(chartTooltipContent, placementNode);
});
BasicChartTooltip.displayName = "BasicChartTooltip";
