// @dynatrace/strato-components/charts/core/components/selection-area/components/Knob.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Knob_exports = {};
__export(Knob_exports, {
  Knob: () => Knob
});
module.exports = __toCommonJS(Knob_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("@use-gesture/react");
var import_clsx = __toESM(require("clsx"));
var import_constants = require("../../overlays/constants.js");
var import_SelectionArea_css = require("../SelectionArea.sty.js");
const Knob = ({
  y,
  x,
  interactable,
  onKeyDown,
  focusable,
  onDragMove,
  onDragStart,
  onDragEnd,
  edgeId,
  hoveredEdge,
  pointerHandlers
}) => {
  const handlers = (0, import_react.useGesture)(
    {
      ...pointerHandlers,
      onDragStart,
      onDragEnd,
      onDrag: ({ xy }) => {
        onDragMove(xy[0]);
      }
    },
    {
      drag: {
        threshold: 2
      }
    }
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "rect",
      {
        y: y - import_constants.KNOB_HEIGHT / 2,
        x: x - import_constants.KNOB_WIDTH / 2 * 2,
        className: (0, import_clsx.default)(import_SelectionArea_css.transparentDraggingArea),
        style: {
          pointerEvents: interactable ? "all" : "none"
        },
        ...handlers()
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "rect",
      {
        y,
        x,
        rx: import_constants.SELECTED_AREA_RADIUS,
        className: (0, import_clsx.default)(
          import_SelectionArea_css.knob,
          "knob",
          hoveredEdge === edgeId && "knob--hovered"
        ),
        tabIndex: focusable ? 0 : -1,
        onKeyDown,
        "aria-label": edgeId
      }
    )
  ] });
};
