// @dynatrace/strato-components/charts/core/components/selection-area/components/SelectionEdge.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SelectionEdge_exports = {};
__export(SelectionEdge_exports, {
  SelectionEdge: () => SelectionEdge
});
module.exports = __toCommonJS(SelectionEdge_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
var import_Knob = require("./Knob.js");
var import_usePoliteContainer = require("../../../../../overlays/hooks/usePoliteContainer.js");
var import_BottomMagnifier = require("../../crosshair/components/BottomMagnifier.js");
var import_TopMagnifier = require("../../crosshair/components/TopMagnifier.js");
var import_constants = require("../../overlays/constants.js");
const SelectionEdge = ({
  x,
  y,
  focusable,
  interactable,
  onKeyDown,
  onDragMove,
  onDragStart,
  onDragEnd,
  topLabel,
  bottomLabel,
  chartBounds,
  isBottomLabelRotated = false,
  isTopLabelRotated = false,
  edgeId,
  hoveredEdge,
  pointerHandlers
}) => {
  const Portal = (0, import_usePoliteContainer.usePoliteContainer)();
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    !(0, import_lodash_es.isUndefined)(bottomLabel) ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(Portal, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_BottomMagnifier.BottomMagnifier,
      {
        label: bottomLabel,
        mouseX: x,
        offset: 0,
        color: import_colors.default.Border.Primary.Accent,
        chartBounds,
        isRotated: isBottomLabelRotated
      }
    ) }) : null,
    !(0, import_lodash_es.isUndefined)(topLabel) ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(Portal, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_TopMagnifier.TopMagnifier,
      {
        label: topLabel,
        mouseX: x,
        offset: 0,
        color: import_colors.default.Border.Primary.Accent,
        chartBounds,
        isRotated: isTopLabelRotated
      }
    ) }) : null,
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Knob.Knob,
      {
        edgeId,
        hoveredEdge,
        y,
        x: x - import_constants.KNOB_WIDTH / 2,
        interactable,
        focusable,
        onKeyDown,
        onDragMove,
        onDragStart,
        onDragEnd,
        pointerHandlers
      }
    )
  ] });
};
