// @dynatrace/strato-components/esm/charts/core/constants.js
import Colors from "@dynatrace/strato-design-tokens/colors";
const SCALE_FALLBACK_COLOR = Colors.Background.Container.Neutral.Accent;
const SIXTY_FPS = 1e3 / 60;
const DEFAULT_DATA_POINT_RADIUS = 3;
const DEFAULT_FILL_OPACITY = 0.8;
const DEFAULT_HIGHLIGHTED_OPACITY = 0.7;
const DEFAULT_INACTIVE_FILL_OPACITY = 0.1;
const INACTIVE_FILL_OPACITY_BAR_CHART = DEFAULT_INACTIVE_FILL_OPACITY;
const INACTIVE_FILL_OPACITY_AREA_CHART = DEFAULT_INACTIVE_FILL_OPACITY;
const INACTIVE_FILL_OPACITY_LINE_CHART = DEFAULT_INACTIVE_FILL_OPACITY;
const INACTIVE_FILL_OPACITY_PIE_CHART = DEFAULT_INACTIVE_FILL_OPACITY;
const INACTIVE_FILL_OPACITY_PIE_CHART_LABEL = 0.2;
const DEFAULT_OPACITY_BAND_CHART = 0.5;
const INACTIVE_FILL_OPACITY_BAND_CHART = 0.15;
const STROKE_ONLY_OPACITY_BAND_CHART = 0.8;
const INACTIVE_STROKE_ONLY_OPACITY_BAND_CHART = 0.15;
const DEFAULT_STROKE_WIDTH_BAND_CHART = 1;
const defaultRatio = "auto";
const DEFAULT_TRUNCATION_MODE = "middle";
const DEFAULT_VALUE_REPRESENTATION = "absolute";
const DEFAULT_VALUE_ACCESSOR = "value";
const DEFAULT_SERIES_CURVE_SHAPE = "linear";
const DEFAULT_INFINITE_ZOOM = false;
const DEFAULT_SELECT_ENABLED = false;
const LAYOUT_PADDING = 8;
const CATEGORICAL_BARS_INNER_PADDING = 0.25;
const CATEGORICAL_BARS_OUTER_PADDING = 0.15;
const CSS_DOWNLOAD_FILENAME = "data.csv";
const LEGEND_POSITION = Object.freeze({
  Auto: "auto",
  Bottom: "bottom",
  Right: "right"
});
const TRUNCATION_MODE = Object.freeze({
  End: "end",
  Middle: "middle",
  Start: "start"
});
const GRAIL_MIN_DOMAIN = /* @__PURE__ */ new Date("1677-09-21T00:12:43.145224192Z");
const GRAIL_MAX_DOMAIN = /* @__PURE__ */ new Date("2262-04-11T23:47:16.854775807Z");
const SWIPE_RELATIVE_THRESHOLD = 0.15;
const PINCH_RELATIVE_THRESHOLD = 0.15;
const RIGHT_BUTTON_ID = 2;
const MIDDLE_BUTTON_ID = 4;
const ChartOverlaysElevation = {
  crosshair: 1,
  dataPoints: 2,
  brush: 3,
  toolbar: 4
};
const FONT_BASE = 10;
const SUFFIX_RATIO = 0.66;
const LOADING_STATE_OPACITY = 40;
const ORIENTATION = {
  Horizontal: "horizontal",
  Vertical: "vertical"
};
const DEFAULT_CUSTOM_TOOLTIP_CONFIG = {
  hidden: false
};
export {
  CATEGORICAL_BARS_INNER_PADDING,
  CATEGORICAL_BARS_OUTER_PADDING,
  CSS_DOWNLOAD_FILENAME,
  ChartOverlaysElevation,
  DEFAULT_CUSTOM_TOOLTIP_CONFIG,
  DEFAULT_DATA_POINT_RADIUS,
  DEFAULT_FILL_OPACITY,
  DEFAULT_HIGHLIGHTED_OPACITY,
  DEFAULT_INACTIVE_FILL_OPACITY,
  DEFAULT_INFINITE_ZOOM,
  DEFAULT_OPACITY_BAND_CHART,
  DEFAULT_SELECT_ENABLED,
  DEFAULT_SERIES_CURVE_SHAPE,
  DEFAULT_STROKE_WIDTH_BAND_CHART,
  DEFAULT_TRUNCATION_MODE,
  DEFAULT_VALUE_ACCESSOR,
  DEFAULT_VALUE_REPRESENTATION,
  FONT_BASE,
  GRAIL_MAX_DOMAIN,
  GRAIL_MIN_DOMAIN,
  INACTIVE_FILL_OPACITY_AREA_CHART,
  INACTIVE_FILL_OPACITY_BAND_CHART,
  INACTIVE_FILL_OPACITY_BAR_CHART,
  INACTIVE_FILL_OPACITY_LINE_CHART,
  INACTIVE_FILL_OPACITY_PIE_CHART,
  INACTIVE_FILL_OPACITY_PIE_CHART_LABEL,
  INACTIVE_STROKE_ONLY_OPACITY_BAND_CHART,
  LAYOUT_PADDING,
  LEGEND_POSITION,
  LOADING_STATE_OPACITY,
  MIDDLE_BUTTON_ID,
  ORIENTATION,
  PINCH_RELATIVE_THRESHOLD,
  RIGHT_BUTTON_ID,
  SCALE_FALLBACK_COLOR,
  SIXTY_FPS,
  STROKE_ONLY_OPACITY_BAND_CHART,
  SUFFIX_RATIO,
  SWIPE_RELATIVE_THRESHOLD,
  TRUNCATION_MODE,
  defaultRatio
};
