// @dynatrace/strato-components/esm/charts/core/components/legend/range/components/RangeLegendItemMarker/RangeLegendItemMarker.js
import { jsx } from "react/jsx-runtime";
import { useRef } from "react";
import Colors from "@dynatrace/strato-design-tokens/colors";
import { legendMarker } from "./RangeLegendItemMarker.sty.js";
import { ORIENTATION } from "../../../../../constants.js";
import {
  CONTAINER_PADDING,
  DEFAULT_GAP_THICKNESS,
  DEFAULT_ITEM_THICKNESS,
  DEFAULT_SEGMENT_SIZE,
  STROKE_WIDTH
} from "../../constants.js";
import { useOrientation } from "../../context/RangeLegendOrientationContext.js";
import { useScale } from "../../context/RangeLegendScaleContext.js";
import { useRangeLegendHoveredMarkerId } from "../../hooks/store/selectors.js";
const RangeLegendItemMarker = (props) => {
  const { markerValue, position, isReversed, isSinglePointRange } = props;
  const svgRef = useRef(null);
  const hoveredMarkerId = useRangeLegendHoveredMarkerId();
  const { scale } = useScale();
  const isBoundarySegment = scale.range().includes(position);
  const { orientation } = useOrientation();
  const isVertical = orientation === ORIENTATION.Vertical;
  const isHorizontal = orientation === ORIENTATION.Horizontal;
  const itemX = isHorizontal ? position + DEFAULT_GAP_THICKNESS / 2 : 0;
  const itemY = isVertical ? position + DEFAULT_GAP_THICKNESS / 2 : 0;
  const itemWidth = isHorizontal ? DEFAULT_GAP_THICKNESS : DEFAULT_ITEM_THICKNESS;
  const itemHeight = isVertical ? DEFAULT_GAP_THICKNESS : DEFAULT_ITEM_THICKNESS;
  const defaultTranslateY = `-${DEFAULT_GAP_THICKNESS / 2}`;
  const verticalSinglePointTranslateY = isReversed ? defaultTranslateY : `-${DEFAULT_GAP_THICKNESS * 2}`;
  const singlePointRangeTranslateY = isHorizontal ? DEFAULT_GAP_THICKNESS : verticalSinglePointTranslateY;
  const CSSTranslate = `translateX(${DEFAULT_GAP_THICKNESS}px) translateY(${isSinglePointRange && !isBoundarySegment ? singlePointRangeTranslateY : defaultTranslateY}px)`;
  const ITEM_X = {
    [ORIENTATION.Vertical]: () => itemX - itemWidth + STROKE_WIDTH,
    [ORIENTATION.Horizontal]: () => {
      if (isSinglePointRange && !isBoundarySegment) {
        return itemX + CONTAINER_PADDING - (DEFAULT_SEGMENT_SIZE + STROKE_WIDTH * 2) / 2;
      }
      return itemX + CONTAINER_PADDING - STROKE_WIDTH * 2;
    }
  };
  const ITEM_Y = {
    [ORIENTATION.Vertical]: () => {
      if (isSinglePointRange && !isBoundarySegment) {
        return itemY - CONTAINER_PADDING - itemHeight / 2;
      }
      return itemY - CONTAINER_PADDING + STROKE_WIDTH;
    },
    [ORIENTATION.Horizontal]: () => {
      return itemY - itemHeight + STROKE_WIDTH;
    }
  };
  return /* @__PURE__ */ jsx(
    "svg",
    {
      width: "48",
      height: "13",
      viewBox: "0 0 48 13",
      fill: "none",
      ref: svgRef,
      x: ITEM_X[orientation](),
      y: ITEM_Y[orientation](),
      className: legendMarker({
        isMarkerShown: hoveredMarkerId === markerValue
      }),
      xmlns: "http://www.w3.org/2000/svg",
      role: "presentation",
      children: /* @__PURE__ */ jsx(
        "path",
        {
          d: "M7.27736 7.08243L7.90139 6.66639L7.27734 6.25038L1.27734 2.25065L0.5 1.73246L0.5 2.66668L0.5 10.6667L0.5 11.601L1.27736 11.0827L7.27736 7.08243Z",
          fill: Colors.Icon.Neutral.Default,
          strokeWidth: STROKE_WIDTH,
          stroke: Colors.Background.Base.Default,
          style: {
            transform: isHorizontal ? `rotate(90deg) ${CSSTranslate}` : CSSTranslate
          }
        }
      )
    }
  );
};
export {
  RangeLegendItemMarker
};
