// @dynatrace/strato-components/esm/charts/timeseries/components/thresholds/ThresholdIndicators.js
import { Fragment, jsx } from "react/jsx-runtime";
import {
  thresholdPlotLine,
  thresholdPlotRange
} from "./ThresholdIndicators.sty.js";
import { useGridPlotSize } from "../../../core/components/chart-grid/hooks/useGridPlotSize.js";
import {
  ThresholdIndicator
} from "../../../core/components/threshold-bar/indicators/ThresholdIndicator.js";
import {
  isDynamic,
  isStatic,
  isStaticRange,
  isDynamicRange
} from "../../../core/types/type-guards.js";
import { transformThresholdIntoIndicator } from "../../../core/utils/thresholds-to-indicator.js";
import { useThresholdsState } from "../../hooks/store/selectors.js";
import { useThresholdConfig } from "../../hooks/useThresholdConfig.js";
import { useXScales } from "../../hooks/useXScale.js";
import { useYScales } from "../../hooks/useYScales.js";
function getThresholdIndicatorType(data, strokeOnly) {
  const isRange = isStaticRange(data) || isDynamicRange(data);
  if (isRange && !strokeOnly) {
    return "range";
  }
  return "line";
}
const ThresholdIndicators = ({
  indicatorType
}) => {
  const { left, right } = useYScales();
  const { currentScale: xScale } = useXScales();
  const { left: leftThresholds, right: rightThresholds } = useThresholdConfig();
  const { width } = useGridPlotSize();
  const { hoveredThresholdId } = useThresholdsState();
  const thresholds = [
    ...leftThresholds ?? [],
    ...rightThresholds ?? []
  ];
  return /* @__PURE__ */ jsx(Fragment, { children: thresholds.map(({ color, data, strokeOnly, position, id }) => {
    if (getThresholdIndicatorType(data, strokeOnly) !== indicatorType) {
      return null;
    }
    const yScale = position === "left" ? left : right;
    if (!yScale) {
      return null;
    }
    const indicatorData = transformThresholdIntoIndicator(data, width, yScale.scale, xScale);
    const isStrokeOnly = isStatic(data) || isDynamic(data) || strokeOnly;
    return /* @__PURE__ */ jsx(
      "svg",
      {
        className: indicatorType === "range" ? thresholdPlotRange : thresholdPlotLine,
        children: /* @__PURE__ */ jsx(
          ThresholdIndicator,
          {
            data: indicatorData,
            color,
            strokeOnly: isStrokeOnly,
            isActive: hoveredThresholdId === id
          }
        )
      },
      `threshold-indicator-${id}`
    );
  }) });
};
export {
  ThresholdIndicators
};
