// @dynatrace/strato-components/esm/charts/pie/components/PieChartToolbar/PieChartToolbar.js
import { jsx } from "react/jsx-runtime";
import { DownloadDataButton } from "./components/DownloadDataButton.js";
import { useBehavioralTrackingProps as _useBehavioralTrackingProps } from "../../../../core/hooks/useBehavioralTrackingProps.js";
import { useTooltipReducer } from "../../../core/components/overlay-tooltip/useTooltipReducer.js";
import { ChartToolbarRenderer } from "../../../core/components/toolbar/ChartToolbarRenderer.js";
import { usePieChartRawData } from "../../context/PieChartRawDataContext.js";
import { useToolbarConfig } from "../../context/ToolbarContext.js";
import { useOverlayState } from "../../hooks/store/selectors.js";
import { zIndicesPieAreas } from "../../types/default.js";
import { buildAndDownloadCsv } from "../../utils/build-and-download-csv.js";
const defaultToolbarPlacement = "top-right";
const PieChartToolbar = () => {
  const dispatch = useTooltipReducer();
  const toolbarConfig = useToolbarConfig();
  const toolbarPlacement = toolbarConfig?.placement ?? defaultToolbarPlacement;
  const { mouseInBounds } = useOverlayState();
  const isDownloadEnabled = toolbarConfig?.downloadData?.enabled ?? false;
  const rawData = usePieChartRawData();
  const handleDownloadDataToolbar = () => buildAndDownloadCsv(rawData);
  const showDefaultToolbarDownloadButton = isDownloadEnabled && !toolbarConfig?.intents;
  const showMenuToolbarButton = isDownloadEnabled && !!toolbarConfig?.intents?.length;
  const initialA11yState = showDefaultToolbarDownloadButton ? [{ download: 0 }] : [{ intent: 0 }];
  const behaviorTrackingProps = _useBehavioralTrackingProps(
    toolbarConfig ?? {}
  )[0];
  return /* @__PURE__ */ jsx(
    ChartToolbarRenderer,
    {
      visible: mouseInBounds,
      placement: toolbarPlacement,
      collapsed: toolbarConfig?.collapsed,
      draggable: true,
      zIndex: zIndicesPieAreas.toolbar,
      initialA11yState,
      onMouseEnter: dispatch({ type: "RESET_TOOLTIP" }),
      menuDisplayMode: "never",
      isDownloadEnabled: showMenuToolbarButton,
      intents: toolbarConfig?.intents,
      handleDownloadDataToolbar,
      removeDragHandle: true,
      ...behaviorTrackingProps,
      children: showDefaultToolbarDownloadButton && /* @__PURE__ */ jsx(DownloadDataButton, {})
    }
  );
};
export {
  PieChartToolbar
};
