// @dynatrace/strato-components/esm/charts/core/components/tooltip/InteractiveChartTooltip/components/BasicChartTooltip/BasicChartTooltip.js
import { jsx } from "react/jsx-runtime";
import {
  forwardRef
} from "react";
import { createPortal } from "react-dom";
import { chartTooltip, wrapper } from "./BasicChartTooltip.sty.js";
import { useOverlayStacking } from "../../../../../../../core/hooks/useOverlayStacking.js";
import { OverlayStackingProvider } from "../../../../../../../core/providers/OverlayStackingProvider.js";
import { useDataDtrumContext } from "../../../../../context/DataDtrumContext.js";
const BasicChartTooltip = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const {
    placement = "left",
    children,
    chartRef,
    onMouseLeave,
    pinned = false
  } = props;
  const { dataDtrumMask, dataDtrumAllow } = useDataDtrumContext();
  const overlayStacking = useOverlayStacking(0, "overlay");
  const chartTooltipContent = /* @__PURE__ */ jsx(
    "div",
    {
      className: wrapper({ isPinned: pinned }),
      style: { zIndex: overlayStacking.style.zIndex },
      "data-testid": "chart-tooltip",
      role: "tooltip",
      ref: forwardedRef,
      "data-placement": placement,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      onMouseLeave,
      children: /* @__PURE__ */ jsx(OverlayStackingProvider, { children: /* @__PURE__ */ jsx("div", { className: chartTooltip, children }) })
    }
  );
  const placementNode = chartRef ? chartRef.current : document.body;
  return createPortal(chartTooltipContent, placementNode);
});
BasicChartTooltip.displayName = "BasicChartTooltip";
export {
  BasicChartTooltip
};
