// @dynatrace/strato-components/charts/core/components/overlay-tooltip/contexts/TooltipContext.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TooltipContext_exports = {};
__export(TooltipContext_exports, {
  TooltipContext: () => TooltipContext,
  TooltipProvider: () => TooltipProvider,
  useTooltip: () => useTooltip
});
module.exports = __toCommonJS(TooltipContext_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_store = require("../../../../../core/store/store.js");
var import_utils = require("../utils.js");
const initialState = () => ({
  pinned: false,
  inBounds: false,
  activeSeries: null,
  absoluteX: null,
  absoluteY: null,
  relativeX: null,
  relativeY: null,
  percentX: null,
  percentY: null,
  data: (0, import_utils.emptyData)()
});
const TooltipContext = (0, import_react.createContext)({});
const useTooltip = () => {
  const tooltip = (0, import_react.useContext)(TooltipContext);
  if (!tooltip) {
    throw new Error("useTooltip not used within TooltipContext provider.");
  }
  return tooltip;
};
function TooltipProvider(props) {
  const store = (0, import_react.useRef)();
  if (props.persistedStore) {
    store.current ||= (0, import_store.createStore)(initialState());
  } else {
    store.current = (0, import_store.createStore)(initialState());
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(TooltipContext.Provider, { value: store.current, children: props.children });
}
