// @dynatrace/strato-components/esm/tables/DataTable/constants.js
const DEFAULT_DATATABLE_CONFIG = {
  fullWidth: false,
  fullHeight: false,
  lineWrap: false,
  selectableRows: void 0,
  resizable: false,
  sortable: false,
  pagination: {},
  variant: {
    contained: true,
    fontStyle: "text",
    headers: "default",
    rowDensity: "default",
    rowSeparation: "horizontalDividers",
    verticalDividers: false
  },
  columnVisibility: {},
  rowThresholds: []
};
const DATA_TABLE_VISUAL_VARIANT_CONFIG_DEFAULT = {
  rowSeparation: "horizontalDividers",
  verticalDividers: false,
  contained: true,
  rowDensity: "default",
  fontStyle: "text",
  headers: "default",
  verticalAlignment: "top"
};
const DATA_TABLE_DEFAULT_PAGINATION_PAGE_INDEX = 0;
const DATA_TABLE_DEFAULT_PAGE_SIZE_OPTIONS = [
  10,
  20,
  50,
  100,
  250,
  500,
  1e3
];
const DATA_TABLE_DEFAULT_PAGE_SIZE_OPTION = DATA_TABLE_DEFAULT_PAGE_SIZE_OPTIONS[3];
const DATA_TABLE_MIN_WIDTH = 120;
const DATA_TABLE_EMPTY_MIN_WIDTH = 200;
const DATA_TABLE_BORDER_WIDTH = 2;
const DATA_TABLE_BORDER_WIDTH_VISUAL_REFRESH = 1;
const DATA_TABLE_DEFAULT_MIN_COLUMN_WIDTH = 30;
const DATA_TABLE_DEFAULT_COLUMN_WIDTH = 150;
const DATA_TABLE_COLUMN_RESIZE_HANDLE_WIDTH = 3;
const DATA_TABLE_BUILTIN_COLUMN_MIN_WIDTH = 24;
const DATA_TABLE_SUB_ROW_SPACER = 16;
const DATA_TABLE_SUB_ROW_INDENT_GRID = DATA_TABLE_SUB_ROW_SPACER;
const DATA_TABLE_THRESHOLD_PILL_SIZE = 4;
const DATA_TABLE_ACTION_BUTTON_WIDTH = 24;
const DATA_TABLE_NARROW_HEADER_THRESHOLDS = {
  condensed: 28 + 6,
  default: 34 + 12,
  comfortable: 38 + 16
};
const DATA_TABLE_HEADER_SORTED_OFFSET_MARGIN = -22;
const DATA_TABLE_HEADER_MULTISORT_OFFSET_MARGIN = -30;
const DATA_TABLE_HEADER_ACTION_OFFSET_MARGIN = -22;
const DATA_TABLE_HEADER_SORTED_ACTION_OFFSET_MARGIN = DATA_TABLE_HEADER_SORTED_OFFSET_MARGIN + DATA_TABLE_HEADER_ACTION_OFFSET_MARGIN;
const DATA_TABLE_HEADER_MULTISORT_ACTION_OFFSET_MARGIN = DATA_TABLE_HEADER_MULTISORT_OFFSET_MARGIN + DATA_TABLE_HEADER_ACTION_OFFSET_MARGIN;
const DATA_TABLE_ROW_HEIGHT_MAP = {
  condensed: 24,
  default: 36,
  comfortable: 44
};
const DATA_TABLE_CELL_VERTICAL_PADDING_MAP = {
  condensed: 2,
  default: 8,
  comfortable: 12
};
const DATA_TABLE_SUB_ROW_DENSITY_PADDING_MAP = DATA_TABLE_CELL_VERTICAL_PADDING_MAP;
const DATA_TABLE_CELL_HORIZONTAL_PADDING_MAP = {
  condensed: 6,
  default: 12,
  comfortable: 16
};
const DATA_TABLE_ROW_BORDER = 1;
const DATA_TABLE_HORIZONTAL_DIVIDER_HEIGHT = 1;
const DATA_TABLE_DEFAULT_MIN_ROW_HEIGHT = DATA_TABLE_ROW_HEIGHT_MAP.condensed;
const DATA_TABLE_HORIZONTAL_SCROLL_SHADOW_WIDTH = 50;
const DATA_TABLE_ZINDEX_CONFIG = {
  DEFAULT: 1,
  STICKY_LEFT: 2,
  STICKY_RIGHT: 3,
  // must be higher than left pinned columns to ensure deterministic overlap
  PINNED_RESIZE_HANDLE: 3,
  OVERLAY: 4,
  HEADER: 5,
  STICKY_SCROLL_SHADOW: 6
};
const DATA_TABLE_ROW_DRAG_COLUMN_ID = "DataTableRowDragColumn";
const DATA_TABLE_ROW_SELECTION_COLUMN_ID = "DataTableRowSelectionColumn";
const DATA_TABLE_ROW_DETAILS_COLUMN_ID = "DataTableRowDetailsColumn";
const DATA_TABLE_ROW_ACTIONS_COLUMN_ID = "DataTableRowActions";
const DATA_TABLE_SPACER_COLUMN_ID = "DataTableSpacerColumn";
const DATA_TABLE_BUILT_IN_COLUMN_IDS = /* @__PURE__ */ new Set([
  DATA_TABLE_ROW_DRAG_COLUMN_ID,
  DATA_TABLE_ROW_SELECTION_COLUMN_ID,
  DATA_TABLE_ROW_DETAILS_COLUMN_ID,
  DATA_TABLE_ROW_ACTIONS_COLUMN_ID,
  DATA_TABLE_SPACER_COLUMN_ID
]);
const DATA_TABLE_DETAILS_ROW_IDENTIFIER = /* @__PURE__ */ Symbol(
  "DETAILS_ROW_IDENTIFIER"
);
const DATA_TABLE_ROW_EXPANDED_ID_SUFFIX = "-expanded";
const DATA_TABLE_PINNING_SHADOW_CELL_ROW_ID_SUFFIX = "-shadow-cell";
const DATA_TABLE_DETAILS_ROW_PARITY = /* @__PURE__ */ Symbol("DETAILS_ROW_PARITY");
const DATA_TABLE_ACTION_FOR_ALL_COLUMNS = /* @__PURE__ */ Symbol(
  "Action applies to all columns"
);
const DATA_TABLE_DOWNLOAD_CSV_FILE_NAME = "table-data.csv";
const DATA_TABLE_INDEX_COLUMN_HEADER = "index";
const DATA_TABLE_RESIZING_LEFT_POSITION_CSS_VAR = "--resizing-leftPosition";
const DATA_TABLE_RESIZING_RIGHT_POSITION_CSS_VAR = "--resizing-rightPosition";
const DATA_TABLE_STRATO_RESIZING_SIZE_CSS_VAR = "--strato-resizing-size";
const DATA_TABLE_DND_PLACEHOLDER_DIVIDER_CSS_VAR = "--column-dnd-placeholder-divider-top";
const DATA_TABLE_DND_NON_TARGET_OPACITY = 0.2;
const DATA_TABLE_VIRTUALIZER_ROW_OVERSCAN = 20;
const DATA_TABLE_VIRTUALIZER_COLUMN_OVERSCAN = 5;
const DATA_TABLE_DEFAULT_SCROLL_OPTIONS = {
  align: "start",
  behavior: "auto"
};
const DATA_TABLE_DEFAULT_ICON_SIZE = 20;
const DATA_TABLE_DEFAULT_ICON_PADDING = 2;
const DATA_TABLE_SUB_ROW_EXPANDER_ICON_SIZE = DATA_TABLE_DEFAULT_ICON_SIZE;
const DATA_TABLE_DEFAULT_HORIZONTAL_ALIGNMENT = "left";
export {
  DATA_TABLE_ACTION_BUTTON_WIDTH,
  DATA_TABLE_ACTION_FOR_ALL_COLUMNS,
  DATA_TABLE_BORDER_WIDTH,
  DATA_TABLE_BORDER_WIDTH_VISUAL_REFRESH,
  DATA_TABLE_BUILTIN_COLUMN_MIN_WIDTH,
  DATA_TABLE_BUILT_IN_COLUMN_IDS,
  DATA_TABLE_CELL_HORIZONTAL_PADDING_MAP,
  DATA_TABLE_CELL_VERTICAL_PADDING_MAP,
  DATA_TABLE_COLUMN_RESIZE_HANDLE_WIDTH,
  DATA_TABLE_DEFAULT_COLUMN_WIDTH,
  DATA_TABLE_DEFAULT_HORIZONTAL_ALIGNMENT,
  DATA_TABLE_DEFAULT_ICON_PADDING,
  DATA_TABLE_DEFAULT_ICON_SIZE,
  DATA_TABLE_DEFAULT_MIN_COLUMN_WIDTH,
  DATA_TABLE_DEFAULT_MIN_ROW_HEIGHT,
  DATA_TABLE_DEFAULT_PAGE_SIZE_OPTION,
  DATA_TABLE_DEFAULT_PAGE_SIZE_OPTIONS,
  DATA_TABLE_DEFAULT_PAGINATION_PAGE_INDEX,
  DATA_TABLE_DEFAULT_SCROLL_OPTIONS,
  DATA_TABLE_DETAILS_ROW_IDENTIFIER,
  DATA_TABLE_DETAILS_ROW_PARITY,
  DATA_TABLE_DND_NON_TARGET_OPACITY,
  DATA_TABLE_DND_PLACEHOLDER_DIVIDER_CSS_VAR,
  DATA_TABLE_DOWNLOAD_CSV_FILE_NAME,
  DATA_TABLE_EMPTY_MIN_WIDTH,
  DATA_TABLE_HEADER_ACTION_OFFSET_MARGIN,
  DATA_TABLE_HEADER_MULTISORT_ACTION_OFFSET_MARGIN,
  DATA_TABLE_HEADER_MULTISORT_OFFSET_MARGIN,
  DATA_TABLE_HEADER_SORTED_ACTION_OFFSET_MARGIN,
  DATA_TABLE_HEADER_SORTED_OFFSET_MARGIN,
  DATA_TABLE_HORIZONTAL_DIVIDER_HEIGHT,
  DATA_TABLE_HORIZONTAL_SCROLL_SHADOW_WIDTH,
  DATA_TABLE_INDEX_COLUMN_HEADER,
  DATA_TABLE_MIN_WIDTH,
  DATA_TABLE_NARROW_HEADER_THRESHOLDS,
  DATA_TABLE_PINNING_SHADOW_CELL_ROW_ID_SUFFIX,
  DATA_TABLE_RESIZING_LEFT_POSITION_CSS_VAR,
  DATA_TABLE_RESIZING_RIGHT_POSITION_CSS_VAR,
  DATA_TABLE_ROW_ACTIONS_COLUMN_ID,
  DATA_TABLE_ROW_BORDER,
  DATA_TABLE_ROW_DETAILS_COLUMN_ID,
  DATA_TABLE_ROW_DRAG_COLUMN_ID,
  DATA_TABLE_ROW_EXPANDED_ID_SUFFIX,
  DATA_TABLE_ROW_HEIGHT_MAP,
  DATA_TABLE_ROW_SELECTION_COLUMN_ID,
  DATA_TABLE_SPACER_COLUMN_ID,
  DATA_TABLE_STRATO_RESIZING_SIZE_CSS_VAR,
  DATA_TABLE_SUB_ROW_DENSITY_PADDING_MAP,
  DATA_TABLE_SUB_ROW_EXPANDER_ICON_SIZE,
  DATA_TABLE_SUB_ROW_INDENT_GRID,
  DATA_TABLE_SUB_ROW_SPACER,
  DATA_TABLE_THRESHOLD_PILL_SIZE,
  DATA_TABLE_VIRTUALIZER_COLUMN_OVERSCAN,
  DATA_TABLE_VIRTUALIZER_ROW_OVERSCAN,
  DATA_TABLE_VISUAL_VARIANT_CONFIG_DEFAULT,
  DATA_TABLE_ZINDEX_CONFIG,
  DEFAULT_DATATABLE_CONFIG
};
