// @dynatrace/strato-components/charts/xy-chart/components/tooltip/XYChartTooltipTemplate/XYChartTooltipTemplate.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var XYChartTooltipTemplate_exports = {};
__export(XYChartTooltipTemplate_exports, {
  XYChartTooltipTemplate: () => XYChartTooltipTemplate
});
module.exports = __toCommonJS(XYChartTooltipTemplate_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_XYChartTooltipTemplate_css = require("./XYChartTooltipTemplate.sty.js");
var import_Flex = require("../../../../../layouts/flex/Flex.js");
var import_useChartActionsMenu = require("../../../../core/components/series-action/hooks/useChartActionsMenu.js");
var import_ChartTooltip = require("../../../../core/components/tooltip/ChartTooltip/ChartTooltip.js");
const XYChartTooltipTemplate = (0, import_react.forwardRef)(({ position, pinned, children }, ref) => {
  const {
    selectedItem: selectedItemId,
    actionsMenuRef,
    updateSelectedItem,
    onLeave
  } = (0, import_useChartActionsMenu.useChartActionsMenu)();
  if (!position) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ChartTooltip.ChartTooltip,
    {
      ref,
      position,
      pinned,
      className: import_XYChartTooltipTemplate_css.body,
      updatePositionWhenPinned: true,
      onMouseLeave: onLeave,
      actionSelectionValue: {
        selectedItemId,
        actionsMenuRef,
        updateSelectedItem
      },
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Flex.Flex, { flexDirection: "column", gap: 4, children })
    }
  );
});
XYChartTooltipTemplate.displayName = "XYChartTooltipTemplate";
