// @dynatrace/strato-components/charts/core/components/overlay-tooltip/utils/build-content.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var build_content_exports = {};
__export(build_content_exports, {
  buildContent: () => buildContent
});
module.exports = __toCommonJS(build_content_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_is_slot = require("../../../../../core/utils/is-slot.js");
var import_Footer = require("../components/Footer.js");
var import_Action = require("../components/global-actions/Action.js");
var import_ActionsPlaceholder = require("../components/global-actions/ActionsPlaceholder.js");
var import_Header = require("../components/header/Header.js");
var import_Series = require("../components/series/Series.js");
var import_SeriesHeader = require("../components/series/SeriesHeader.js");
var import_SeriesItem = require("../components/series/SeriesItem.js");
var import_constants = require("../constants.js");
const VALID_TRUNCATION_MODES = ["start", "middle", "end"];
const isValidTruncationMode = (value) => {
  return typeof value === "string" && VALID_TRUNCATION_MODES.includes(value);
};
const getTruncationMode = (userTruncationMode, defaultConfigTruncationMode) => {
  if (isValidTruncationMode(userTruncationMode)) {
    return userTruncationMode;
  }
  if (isValidTruncationMode(defaultConfigTruncationMode)) {
    return defaultConfigTruncationMode;
  }
  return import_constants.DEFAULT_TRUNCATION_MODE;
};
function buildContent(children, defaultConfig, args) {
  const slots = {
    Header: null,
    Footer: null,
    Series: null,
    SeriesHeader: null,
    SeriesItem: null,
    ActionsPlaceholder: null,
    GlobalActions: []
  };
  let unpinnedPlaceholderText;
  let defaultActionsConfig = defaultConfig?.defaultActionsConfig;
  if (!args.customTooltip) {
    import_react.Children.forEach(children, (child) => {
      if ((0, import_is_slot.isSlot)(child, import_Header.Header)) {
        slots.Header = child;
      } else if ((0, import_is_slot.isSlot)(child, import_Footer.Footer)) {
        slots.Footer = child;
      } else if ((0, import_is_slot.isSlot)(child, import_SeriesHeader.SeriesHeader)) {
        slots.SeriesHeader = child;
        defaultActionsConfig = (0, import_lodash_es.clone)(
          defaultConfig?.defaultActionsConfig
        );
        const defaultNameAction = defaultActionsConfig?.get("name");
        if (child.props.seriesLabel && defaultNameAction) {
          defaultActionsConfig?.set("name", {
            ...defaultNameAction,
            customLabel: child.props.seriesLabel
          });
        }
        const defaultValueAction = defaultActionsConfig?.get("value");
        if (child.props.valueLabel && defaultValueAction) {
          defaultActionsConfig?.set("value", {
            ...defaultValueAction,
            customLabel: child.props.valueLabel
          });
        }
      } else if ((0, import_is_slot.isSlot)(child, import_Series.Series)) {
        slots.Series = child;
      } else if ((0, import_is_slot.isSlot)(child, import_SeriesItem.SeriesItem)) {
        slots.SeriesItem = child;
      } else if ((0, import_is_slot.isSlot)(child, import_ActionsPlaceholder.ActionsPlaceholder)) {
        unpinnedPlaceholderText = child.props.text;
      } else if ((0, import_is_slot.isSlot)(child, import_Action.Action)) {
        slots.GlobalActions.push(child);
      }
    });
  }
  return {
    content: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react.Fragment, { children: slots.Header }, "header"),
      slots.SeriesItem ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react.Fragment, { children: slots.SeriesHeader || /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_SeriesHeader.SeriesHeader, {}) }, "seriesHeader") : null,
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react.Fragment, { children: slots.Series || slots.SeriesItem }, "series"),
      slots.Footer ?? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_Footer.Footer,
        {
          actions: slots.GlobalActions,
          unpinnedPlaceholderText
        },
        "footer"
      )
    ],
    config: {
      seriesHeader: slots.SeriesHeader || /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_SeriesHeader.SeriesHeader, {}),
      displayHeaderDivider: !!(slots.Header && (slots.Series || slots.SeriesItem)),
      symbolPosition: args.userSymbolPosition || defaultConfig?.symbolPosition || "left",
      truncationMode: getTruncationMode(
        args.userTruncationMode,
        defaultConfig?.truncationMode
      ),
      defaultActionsConfig,
      variant: args.customTooltip || defaultConfig?.sharedVariant && slots.Series?.props.variant === "shared" ? "shared" : "single"
    }
  };
}
