// @dynatrace/strato-components/charts/xy-chart/components/axes/hooks/useXTickOffset.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useXTickOffset_exports = {};
__export(useXTickOffset_exports, {
  useXTickOffset: () => useXTickOffset
});
module.exports = __toCommonJS(useXTickOffset_exports);
var import_constants = require("../../../../core/components/axes/constants.js");
function useXTickOffset(textWidth, textHeight, renderProps, centerTick, isTickRotated, position, tickNameHasEmoji) {
  const yOffset = position === "top" && !isTickRotated ? import_constants.TOP_AXIS_LINE_OFFSET : import_constants.BOTTOM_AXIS_LINE_OFFSET;
  const size = isTickRotated ? textHeight + 2 : textWidth;
  const xOffset = centerTick ? size / 2 : renderProps.textAnchor === "start" ? 0 : renderProps.textAnchor === "end" ? size : size / 2;
  const { x, y, ...restProps } = renderProps;
  const TICK_LINE_HEIGHT_VALIDATED = tickNameHasEmoji ? import_constants.TICK_LINE_HEIGHT + import_constants.TICK_LINE_ICON_OFFSET : import_constants.TICK_LINE_HEIGHT;
  const style = {
    ...restProps,
    width: textWidth,
    height: TICK_LINE_HEIGHT_VALIDATED,
    whiteSpace: "nowrap",
    textOverflow: "ellipsis",
    display: "inline-block"
  };
  if (isTickRotated) {
    style.transform = "rotate(-90deg)";
  }
  return { xOffset, yOffset, style };
}
