// @dynatrace/strato-components/charts/xy-chart/components/axes/hooks/useYTickOffset.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useYTickOffset_exports = {};
__export(useYTickOffset_exports, {
  useYTickOffset: () => useYTickOffset
});
module.exports = __toCommonJS(useYTickOffset_exports);
var import_typography = __toESM(require("@dynatrace/strato-design-tokens/typography"));
var import_constants = require("../../../../core/components/axes/constants.js");
const SMALL_GAP_OFFSET = 1;
const START_ANCHOR_OFFSET = 3;
function getYTickOffset(centerTick, renderProps, tickNameHasEmoji) {
  const TICK_LINE_HEIGHT_VALIDATED = tickNameHasEmoji ? import_constants.TICK_LINE_HEIGHT + import_constants.TICK_LINE_ICON_OFFSET : import_constants.TICK_LINE_HEIGHT;
  if (centerTick) {
    return TICK_LINE_HEIGHT_VALIDATED / 2 + SMALL_GAP_OFFSET;
  }
  if (renderProps.verticalAnchor === "start") {
    return START_ANCHOR_OFFSET;
  }
  if (renderProps.verticalAnchor === "end") {
    return TICK_LINE_HEIGHT_VALIDATED - SMALL_GAP_OFFSET;
  }
  return TICK_LINE_HEIGHT_VALIDATED / 2 + SMALL_GAP_OFFSET;
}
function useYTickOffset(position, textWidth, renderProps, centerTick, tickNameHasEmoji) {
  const WIDTH = Math.ceil(textWidth);
  const yOffset = getYTickOffset(centerTick, renderProps, tickNameHasEmoji);
  const xOffset = position === "left" ? textWidth + import_constants.Y_AXES_TICK_LINE_OFFSET - SMALL_GAP_OFFSET : -import_constants.Y_AXES_TICK_LINE_OFFSET + SMALL_GAP_OFFSET;
  const { x, y, ...restProps } = renderProps;
  const style = {
    ...restProps,
    textAlign: position === "left" ? "end" : "start",
    lineHeight: import_typography.default.Text.Base.Default.LineHeight,
    width: WIDTH,
    height: import_constants.TICK_LINE_HEIGHT,
    overflow: "visible"
  };
  return { xOffset, yOffset, style };
}
