// @dynatrace/strato-components/charts/gantt-chart/components/tooltip/GanttChartTooltipContent.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var GanttChartTooltipContent_exports = {};
__export(GanttChartTooltipContent_exports, {
  GanttChartTooltipContent: () => GanttChartTooltipContent
});
module.exports = __toCommonJS(GanttChartTooltipContent_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_spacings = __toESM(require("@dynatrace/strato-design-tokens/spacings"));
var import_GanttChartTooltipBody = require("./GanttChartTooltipBody.js");
var import_GanttChartTooltipHeader = require("./GanttChartTooltipHeader.js");
var import_Divider = require("../../../../layouts/divider/Divider.js");
var import_ChartTooltip = require("../../../core/components/tooltip/ChartTooltip/ChartTooltip.js");
var import_GanttSegmentsContext = require("../../context/GanttSegmentsContext.js");
var import_use_set_tooltip_state = require("../../hooks/use-set-tooltip-state.js");
const GanttChartTooltipContent = () => {
  const [tooltipContent, setTooltipContent] = (0, import_react.useState)();
  const isTooltipPinned = (0, import_use_set_tooltip_state.useTooltipPinnedStatus)();
  const storedTooltipContent = (0, import_use_set_tooltip_state.useTooltipContent)();
  const {
    config: { xAxisType, tooltip, formatter }
  } = (0, import_GanttSegmentsContext.useGanttSegmentsContext)();
  const isDate = xAxisType === "time";
  (0, import_react.useEffect)(() => {
    if (!isTooltipPinned && storedTooltipContent) {
      setTooltipContent(storedTooltipContent);
    }
  }, [storedTooltipContent, isTooltipPinned]);
  if (!tooltipContent) {
    return null;
  }
  const { row, segment, parent } = tooltipContent;
  if (tooltip !== void 0) {
    return tooltip(segment, row, parent);
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_GanttChartTooltipHeader.GanttChartTooltipHeader,
      {
        segment: tooltipContent.segment,
        isDate,
        formatter
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Divider.Divider,
      {
        orientation: "horizontal",
        style: { margin: `${import_spacings.default.Size4} 0` }
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ChartTooltip.ChartTooltip.Body, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_GanttChartTooltipBody.GanttChartTooltipBody,
      {
        entry: tooltipContent.segment,
        row,
        parent,
        isDate
      }
    ) })
  ] });
};
GanttChartTooltipContent["displayName"] = "GanttChartTooltipBody";
