// @dynatrace/strato-components/charts/gantt-chart/components/tooltip/GanttChartTooltip.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var GanttChartTooltip_exports = {};
__export(GanttChartTooltip_exports, {
  GanttChartTooltip: () => GanttChartTooltip
});
module.exports = __toCommonJS(GanttChartTooltip_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_GanttChartTooltip_css = require("./GanttChartTooltip.sty.js");
var import_GanttChartTooltipContent = require("./GanttChartTooltipContent.js");
var import_useChartActionsMenu = require("../../../core/components/series-action/hooks/useChartActionsMenu.js");
var import_ChartTooltip = require("../../../core/components/tooltip/ChartTooltip/ChartTooltip.js");
var import_defaults = require("../../constants/defaults.js");
var import_GanttSegmentsContext = require("../../context/GanttSegmentsContext.js");
var import_use_register_tooltip_handler_callbacks = require("../../hooks/use-register-tooltip-handler-callbacks.js");
var import_use_set_tooltip_state = require("../../hooks/use-set-tooltip-state.js");
const GanttChartTooltip = () => {
  const tooltipRef = (0, import_react.useRef)(null);
  const pinned = (0, import_use_set_tooltip_state.useTooltipPinnedStatus)();
  const mousePosition = (0, import_use_set_tooltip_state.useTooltipPosition)();
  const { hovering = true } = (0, import_react.useContext)(import_GanttSegmentsContext.GanttSegmentsContext);
  const {
    selectedItem: selectedItemId,
    actionsMenuRef,
    updateSelectedItem,
    onLeave
  } = (0, import_useChartActionsMenu.useChartActionsMenu)();
  (0, import_use_register_tooltip_handler_callbacks.useRegisterTooltipHandlerCallbacks)(tooltipRef);
  const isTooltipHeightKnown = (0, import_react.useRef)(false);
  if (!mousePosition || !hovering) {
    return null;
  }
  if (!isTooltipHeightKnown.current && tooltipRef.current) {
    isTooltipHeightKnown.current = true;
    return null;
  }
  const position = getTooltipPosition(mousePosition, tooltipRef);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ChartTooltip.ChartTooltip,
    {
      ref: tooltipRef,
      pinned,
      className: import_GanttChartTooltip_css.ganttChartTooltipCSS,
      position,
      onMouseLeave: onLeave,
      actionSelectionValue: {
        selectedItemId,
        actionsMenuRef,
        updateSelectedItem
      },
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_GanttChartTooltipContent.GanttChartTooltipContent, {})
    }
  );
};
function getTooltipPosition(mousePosition, tooltipRef) {
  if (!tooltipRef.current) {
    return {
      x: -9999,
      y: -9999
    };
  }
  const position = {
    x: mousePosition.x,
    y: mousePosition.y + import_defaults.DEFAULT_TOOLTIP_MARGIN
  };
  const tooltipHeight = tooltipRef.current.getBoundingClientRect().height;
  const viewportHeight = window.innerHeight;
  if (position.y + tooltipHeight > viewportHeight) {
    position.y = viewportHeight - tooltipHeight;
  }
  return {
    x: Math.max(0, position.x),
    y: Math.max(0, position.y)
  };
}
