// @dynatrace/strato-components/esm/tables/DataTable/hooks/useColumns.js
import { jsx } from "react/jsx-runtime";
import { useMemo } from "react";
import { makeAccessorFunction } from "./makeAccessorFunction.js";
import { LogContentHighlight } from "../../utils/log-content-highlighting.js";
import { replaceUrlsWithLinks } from "../../utils/replace-urls-with-links.js";
import {
  applyColumnFormatters,
  formatCellValue
} from "../features/CellFormatter/CellFormatter.js";
import {
  applyColumnTypes,
  getAlignmentBasedOnType,
  isDatatableLogContentColumnConfig
} from "../features/ColumnTypes/ColumnTypes.js";
import { SPACER_COLUMN } from "../features/Layout/Layout.js";
import { applySortTypes } from "../features/Sorting/sort-types.js";
function isDataTableGroupColumnDef(columnDef) {
  return columnDef !== null && typeof columnDef === "object" && "columns" in columnDef && Array.isArray(columnDef.columns);
}
function isDataTableDisplayColumnDef(columnDef) {
  return columnDef !== null && typeof columnDef === "object" && "accessor" in columnDef;
}
function splitAccessor(column) {
  if (typeof column.accessor === "string") {
    if (column.accessor.includes(".") || column.accessor.includes("[")) {
      return {
        accessorFn: makeAccessorFunction(column.accessor)
      };
    }
    return {
      accessorKey: column.accessor
    };
  }
  if (typeof column.accessor === "function") {
    return {
      accessorFn: column.accessor
    };
  }
  return {};
}
function createHeaderRenderer(header) {
  if (typeof header === "function") {
    return () => header();
  }
  return header;
}
function createCellRenderer(consumerCustomCell, consumerColumnDef) {
  return ({ getValue, row, column }) => {
    const lineWrap = column.getIsLineWrapped();
    return consumerCustomCell({
      value: getValue(),
      rowIndex: row.index,
      rowData: row.original,
      rowId: row.id,
      format: (value) => formatCellValue(value, consumerColumnDef),
      formatLogContent: (logContent) => {
        const truncationLimit = isDatatableLogContentColumnConfig(
          consumerColumnDef.config
        ) ? consumerColumnDef.config.truncationLimit : void 0;
        return /* @__PURE__ */ jsx(
          LogContentHighlight,
          {
            logContent,
            lineWrap,
            truncationLimit
          }
        );
      },
      detectLinks: replaceUrlsWithLinks,
      isLineWrapped: lineWrap
    });
  };
}
function createDisplayColumnDef(columnDef) {
  const hasFormatter = columnDef.columnType !== "meterbar" && columnDef.columnType !== "sparkline" && columnDef.columnType !== "gantt" && !!columnDef.formatter;
  const composedDef = {
    id: columnDef.id,
    header: createHeaderRenderer(columnDef.header),
    label: columnDef.label,
    width: columnDef.width,
    minWidth: columnDef.minWidth,
    maxWidth: columnDef.maxWidth,
    alignment: columnDef.alignment ?? getAlignmentBasedOnType(columnDef.columnType),
    enableSorting: columnDef.disableSorting ? !columnDef.disableSorting : true,
    sortAccessor: columnDef.sortAccessor,
    sortDescFirst: columnDef.sortDescFirst,
    invertSorting: columnDef.sortInverted,
    enableHiding: !columnDef.disableColumnHiding,
    thresholds: columnDef.thresholds,
    meta: {
      config: columnDef.config,
      hasFormatter,
      hasCustomCellRenderer: columnDef.cell !== void 0,
      hasCustomHeaderRenderer: typeof columnDef.header !== "string"
    },
    ...splitAccessor(columnDef),
    ...applySortTypes(columnDef)
  };
  if (columnDef.columnType) {
    composedDef.columnType = columnDef.columnType;
  }
  if (columnDef.cell) {
    const customCell = columnDef.cell;
    if (typeof customCell === "string") {
      composedDef.cell = customCell;
      return composedDef;
    }
    if (typeof customCell === "function") {
      return {
        ...composedDef,
        cell: createCellRenderer(customCell, columnDef)
      };
    }
    return composedDef;
  }
  return {
    ...composedDef,
    ...applyColumnTypes(columnDef, composedDef),
    ...applyColumnFormatters(columnDef)
  };
}
function useColumns(propColumns) {
  return useMemo(() => {
    const createdColumns = [];
    const propColumnMap = /* @__PURE__ */ new Map();
    for (const firstLevelColumn of propColumns) {
      propColumnMap.set(firstLevelColumn.id, firstLevelColumn);
      if (isDataTableGroupColumnDef(firstLevelColumn)) {
        const groupColumn = {
          id: firstLevelColumn.id,
          header: createHeaderRenderer(firstLevelColumn.header),
          width: firstLevelColumn.width,
          enableSorting: false,
          columns: firstLevelColumn.columns?.map(
            (childColumn) => createDisplayColumnDef(childColumn)
          ),
          alignment: firstLevelColumn.alignment,
          meta: {
            hasCustomHeaderRenderer: typeof firstLevelColumn.header !== "string"
          }
        };
        createdColumns.push(groupColumn);
        for (const secondLevelColumn of firstLevelColumn.columns ?? []) {
          propColumnMap.set(
            secondLevelColumn.id,
            secondLevelColumn
          );
        }
      } else if (isDataTableDisplayColumnDef(firstLevelColumn)) {
        createdColumns.push(createDisplayColumnDef(firstLevelColumn));
      }
    }
    if (createdColumns.length > 0) {
      createdColumns.push(SPACER_COLUMN);
    }
    return [createdColumns, propColumnMap];
  }, [propColumns]);
}
export {
  isDataTableDisplayColumnDef,
  isDataTableGroupColumnDef,
  useColumns
};
