// @dynatrace/strato-components/esm/tables/DataTable/hooks/useDataTableSlots.js
import { Children, useMemo } from "react";
import { isElement } from "react-is";
import { _isEmptyState } from "../components/DataTableEmptyState.js";
import {
  _isDataTableSelectedRowsActions
} from "../components/TableActions/DataTableSelectedRowsActions.js";
import {
  _isTableActions
} from "../components/TableActions/DataTableTableActions.js";
import {
  _isDataTableToolbar
} from "../components/Toolbar/DataTableToolbar.js";
import { _isDataTablePagination } from "../features/Pagination/Pagination.js";
function useDataTableSlots(children) {
  return useMemo(() => {
    const slots = {
      pagination: null,
      toolbarProps: null,
      selectedRowsActions: null,
      emptyState: null,
      tableActions: null
    };
    Children.forEach(children, (child) => {
      if (isElement(child)) {
        if (_isDataTableToolbar(child)) {
          slots.toolbarProps = child.props;
        }
        if (_isTableActions(child)) {
          slots.tableActions = child;
        }
        if (_isEmptyState(child)) {
          slots.emptyState = child;
        }
        if (_isDataTablePagination(child)) {
          slots.pagination = child;
        }
        if (_isDataTableSelectedRowsActions(child)) {
          slots.selectedRowsActions = child;
        }
      }
    });
    return slots;
  }, [children]);
}
export {
  useDataTableSlots
};
