// @dynatrace/strato-components/esm/tables/DataTable/hooks/useHeaderDragAndDrop.js
import { useSortable } from "@dnd-kit/sortable";
import { useRef } from "react";
import { isDragMotion } from "../features/ColumnOrder/column-order-dnd-utils.js";
function useHeaderCellDragAndDrop({
  headerId,
  column,
  columnIndex,
  columnDragAndDropMode
}) {
  const touchStart = useRef();
  const portalRef = useRef(null);
  const { setNodeRef, isDragging, attributes, listeners } = useSortable({
    id: headerId,
    data: {
      portalRef,
      column,
      columnIndex
    }
  });
  const dragAndDropProps = columnDragAndDropMode === "none" ? void 0 : {
    ...attributes,
    draggable: true,
    onDragStart: (event) => listeners?.onMouseDown(event),
    onTouchStart: (event) => {
      touchStart.current = event;
    },
    onTouchMove: (event) => {
      if (isDragMotion(event, touchStart?.current)) {
        listeners?.onTouchStart(event);
      }
    }
  };
  return {
    setNodeRef,
    isDragging,
    portalRef,
    dragAndDropProps
  };
}
export {
  useHeaderCellDragAndDrop
};
