// @dynatrace/strato-components/charts/core/components/interactions-overlay/interactions/pan/dragInteractions.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var dragInteractions_exports = {};
__export(dragInteractions_exports, {
  panDragEnd: () => panDragEnd,
  panDragEndWithMiddleButton: () => panDragEndWithMiddleButton,
  panDragMove: () => panDragMove,
  panDragMoveWithMiddleButton: () => panDragMoveWithMiddleButton,
  panDragStart: () => panDragStart,
  panDragStartWithMiddleButton: () => panDragStartWithMiddleButton
});
module.exports = __toCommonJS(dragInteractions_exports);
var import_lodash_es = require("lodash");
var import_center_position_vertically = require("../../utils/center-position-vertically.js");
const panDragStart = ({
  panDeltaRef,
  onPanStart
}) => {
  onPanStart?.({ triggeredBy: "pointer" });
  panDeltaRef.current = 0;
};
const panDragMove = ({
  event,
  height,
  panDeltaRef,
  onPanChange
}) => {
  const {
    startHoldingAt: startHoldingAtProp,
    currentPosition: currentPositionProp
  } = event;
  const startHoldingAt = (0, import_center_position_vertically.centerPositionVertically)(startHoldingAtProp, height);
  const currentPosition = (0, import_center_position_vertically.centerPositionVertically)(currentPositionProp, height);
  if (!(0, import_lodash_es.isUndefined)(panDeltaRef.current)) {
    const fullInteractionDelta = currentPosition.relative.x - startHoldingAt.relative.x;
    const currentDelta = fullInteractionDelta - panDeltaRef.current;
    panDeltaRef.current = fullInteractionDelta;
    onPanChange?.({ triggeredBy: "pointer", delta: -currentDelta });
  }
};
const panDragEnd = ({
  panDeltaRef,
  onPanEnd
}) => {
  panDeltaRef.current = void 0;
  onPanEnd?.({ triggeredBy: "pointer" });
};
const panDragStartWithMiddleButton = ({
  panDeltaRef,
  onPanStart
}) => {
  onPanStart?.({ triggeredBy: "pointer" });
  panDeltaRef.current = 0;
};
const panDragMoveWithMiddleButton = ({
  event,
  panDeltaRef,
  onPanChange
}) => {
  if (!(0, import_lodash_es.isUndefined)(panDeltaRef.current)) {
    const { currentPosition, startHoldingAt } = event;
    const fullInteractionDelta = currentPosition.relative.x - startHoldingAt.relative.x;
    const currentDelta = fullInteractionDelta - panDeltaRef.current;
    panDeltaRef.current = fullInteractionDelta;
    onPanChange?.({ triggeredBy: "pointer", delta: -currentDelta });
  }
};
const panDragEndWithMiddleButton = ({
  event,
  panDeltaRef,
  lastMousePositionRef,
  onPanEnd
}) => {
  panDeltaRef.current = void 0;
  lastMousePositionRef.current = event.releasedAt;
  onPanEnd?.({ triggeredBy: "pointer" });
};
