// @dynatrace/strato-components/charts/core/components/interactions-overlay/interactions/pan/keyboardInteractions.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var keyboardInteractions_exports = {};
__export(keyboardInteractions_exports, {
  panKeyboardEscape: () => panKeyboardEscape,
  panKeyboardLeft: () => panKeyboardLeft,
  panKeyboardRight: () => panKeyboardRight
});
module.exports = __toCommonJS(keyboardInteractions_exports);
var import_lodash_es = require("lodash");
var import_constants = require("../../constants.js");
const panKeyboardEscape = ({
  selectionArea,
  eventCancelled,
  onSelectEnd,
  lastMousePositionRef,
  pinned,
  onHover,
  onTogglePin,
  onModeChange,
  onPanEnd
}) => {
  onPanEnd?.({ triggeredBy: "keyboard" });
  if (!(0, import_lodash_es.isUndefined)(selectionArea)) {
    eventCancelled.current = true;
    onSelectEnd?.({ cancelled: true });
    if (lastMousePositionRef.current) {
      onHover?.({
        triggeredBy: "pointer",
        position: lastMousePositionRef.current
      });
    }
  }
  if (pinned) {
    onTogglePin?.({
      triggeredBy: "keyboard",
      pin: false,
      position: lastMousePositionRef.current
    });
  }
  onModeChange?.({ mode: "explore", keyboard: false });
};
const panKeyboardLeft = ({
  shiftKey,
  width,
  onPanChange
}) => {
  const factor = shiftKey ? -import_constants.LARGE_STEP_FACTOR : -import_constants.SMALL_STEP_FACTOR;
  const delta = width * factor;
  onPanChange?.({ triggeredBy: "keyboard", delta });
};
const panKeyboardRight = ({
  shiftKey,
  width,
  onPanChange
}) => {
  const factor = shiftKey ? import_constants.LARGE_STEP_FACTOR : import_constants.SMALL_STEP_FACTOR;
  const delta = width * factor;
  onPanChange?.({ triggeredBy: "keyboard", delta });
};
