// @dynatrace/strato-components/tables/DataTable/features/RowSelection/RowSelectionCells.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var RowSelectionCells_exports = {};
__export(RowSelectionCells_exports, {
  RowSelectionCell: () => RowSelectionCell,
  RowSelectionHeader: () => RowSelectionHeader
});
module.exports = __toCommonJS(RowSelectionCells_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_Checkbox = require("../../../../forms/checkbox/Checkbox.js");
var import_Tooltip = require("../../../../overlays/tooltip/Tooltip.js");
var import_DataTableDefaultCell = require("../../components/DataTableDefaultCell.js");
var import_messages = require("../../messages.js");
function RowSelectionHeader(props) {
  const { table } = props;
  const resetBatchSelection = table.resetBatchSelection;
  const value = getCheckboxValue(table);
  const isDisabled = isSelectAllDisabled(table);
  const intl = (0, import_react_intl.useIntl)();
  const checkboxMessage = intl.formatMessage(
    import_messages.dataTableMessages.selectAllTooltip,
    {
      nextSelection: value === false || value === "indeterminate" ? "select" : "deselect",
      scope: table.options.rowSelectionSelectAllMode
    }
  );
  (0, import_react.useEffect)(() => {
    document.addEventListener("click", resetBatchSelection);
    return () => {
      document.removeEventListener("click", resetBatchSelection);
    };
  }, [resetBatchSelection]);
  function toggleSelectAll(event) {
    if (isDisabled) {
      event?.preventDefault();
      return;
    }
    if (table.options.rowSelectionSelectAllMode === "page") {
      if (table.options.rowSelectionLimit !== void 0) {
        table.getToggleAllLimitedPageRowsSelectedHandler()(event);
      } else {
        table.getToggleAllPageRowsSelectedHandler()(event);
      }
    } else if (table.options.rowSelectionSelectAllMode === "all") {
      if (table.options.rowSelectionLimit !== void 0) {
        table.getToggleAllLimitedRowsSelectedHandler()(event);
      } else {
        table.getToggleAllRowsSelectedHandler()(event);
      }
    }
  }
  const selectAllDtProperties = `state:${value === true ? "deselect" : "select"}`;
  const checkbox = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Checkbox.Checkbox,
    {
      value: isDisabled ? false : value,
      "aria-disabled": isDisabled,
      "aria-label": checkboxMessage,
      onChange: (_, event) => toggleSelectAll(event),
      "data-dt-name": "RowSelectDeselectAll",
      "data-dt-properties": selectAllDtProperties
    }
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_DataTableDefaultCell.DataTableDefaultCell, { children: isDisabled ? checkbox : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Tooltip.Tooltip, { text: checkboxMessage, children: checkbox }) });
}
function isSelectAllDisabled(table) {
  const isTableEmpty = table.getRowCount() === 0;
  if (isTableEmpty) {
    return true;
  }
  let isAllRowsSelectionDisabled = false;
  if (table.options.rowSelectionSelectAllMode === "page") {
    isAllRowsSelectionDisabled = table.getIsAllPageRowsSelectionDisabled();
  } else if (table.options.rowSelectionSelectAllMode === "all") {
    isAllRowsSelectionDisabled = table.getIsAllRowsSelectionDisabled();
  }
  return isAllRowsSelectionDisabled;
}
function getCheckboxValue(table) {
  if (table.options.rowSelectionSelectAllMode === "page") {
    if (table.getIsAllPageRowsSelected()) {
      return true;
    }
    if (table.getIsSomePageRowsSelected()) {
      return "indeterminate";
    }
    return false;
  }
  if (table.options.rowSelectionSelectAllMode === "all") {
    if (table.getIsAllRowsSelected()) {
      return true;
    }
    if (table.getIsSomeRowsSelected()) {
      return "indeterminate";
    }
    return false;
  }
  return false;
}
const resolveCheckboxValue = (row) => {
  if (row.subRows.length === 0) {
    return row.getIsSelected();
  }
  if (row.getIsAllSubRowsSelected()) {
    return true;
  }
  if (row.getIsSomeSelected()) {
    return "indeterminate";
  }
  return false;
};
function RowSelectionCell(props) {
  const { row } = props;
  const intl = (0, import_react_intl.useIntl)();
  const isDisabled = row.getShouldDisableRowSelection();
  const rowDtProperties = `state:${resolveCheckboxValue(row) === true ? "deselect" : "select"}`;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_DataTableDefaultCell.DataTableDefaultCell,
    {
      onMouseDown: (event) => {
        document.getSelection()?.removeAllRanges();
        event.preventDefault();
      },
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_Checkbox.Checkbox,
        {
          "aria-label": intl.formatMessage(import_messages.dataTableMessages.selectRowCheckbox, {
            id: row.id
          }),
          value: resolveCheckboxValue(row),
          "aria-disabled": isDisabled,
          onChange: (value, event) => {
            if (isDisabled) {
              event?.preventDefault();
              return;
            }
            row.batchSelect(value, (event?.nativeEvent).shiftKey);
          },
          "data-row-selection-checkbox": true,
          "data-dt-name": "RowSelectDeselect",
          "data-dt-properties": rowDtProperties
        }
      )
    }
  );
}
