// @dynatrace/strato-components/tables/DataTable/features/RowSelection/RowSelection.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var RowSelection_exports = {};
__export(RowSelection_exports, {
  DataTableRowSelection: () => DataTableRowSelection,
  ROW_SELECTION_COLUMN: () => ROW_SELECTION_COLUMN,
  isRowSelectionColumnDef: () => isRowSelectionColumnDef,
  useRowSelection: () => useRowSelection
});
module.exports = __toCommonJS(RowSelection_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_RowSelectionCells = require("./RowSelectionCells.js");
var import_useControllableState = require("../../../../core/hooks/useControllableState.js");
var import_useDataTableConfig = require("../../../hooks/useDataTableConfig.js");
var import_constants = require("../../constants.js");
var import_utils = require("../../hooks/useTable/utils.js");
function isRowSelectionColumnDef(column) {
  return column.id === ROW_SELECTION_COLUMN.id;
}
const ROW_SELECTION_COLUMN = {
  id: import_constants.DATA_TABLE_ROW_SELECTION_COLUMN_ID,
  alignment: "left",
  accessorFn: () => "",
  header: ({ table }) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_RowSelectionCells.RowSelectionHeader, { table }),
  width: "content",
  minWidth: import_constants.DATA_TABLE_BUILTIN_COLUMN_MIN_WIDTH,
  enableSorting: false,
  enableResizing: false,
  enableUserActions: false,
  enableHiding: false,
  meta: {
    builtin: true,
    hasCustomCellRenderer: true,
    hasCustomHeaderRenderer: true,
    builtinPlacement: "prefix",
    cellDeps: (cell) => {
      return [
        cell.row.getIsSelected(),
        cell.row.getIsAllSubRowsSelected(),
        cell.row.getIsSomeSelected(),
        cell.row.getCanSelect(),
        cell.row.getShouldDisableRowSelection()
      ];
    }
  },
  cell: ({ row }) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_RowSelectionCells.RowSelectionCell, { row })
};
const mutateRowIsSelected = (selectedRowIds, id, value, includeChildren, table) => {
  const row = table.getRow(id, true);
  if (!row.getShouldDisableRowSelection()) {
    if (value) {
      if (row.getCanSelect() && !table.getIsRowSelectionLimitReached(selectedRowIds)) {
        selectedRowIds[id] = true;
      }
    } else {
      delete selectedRowIds[id];
    }
  }
  if (includeChildren && row.subRows?.length && row.getCanSelectSubRows() && !table.getIsRowSelectionLimitReached(selectedRowIds)) {
    row.subRows.forEach(
      (row2) => mutateRowIsSelected(
        selectedRowIds,
        row2.id,
        value,
        includeChildren,
        table
      )
    );
  }
};
const DataTableRowSelection = {
  createTable(table) {
    let _batchSelectionRowId = void 0;
    let _previousSelectionRange = void 0;
    table.resetBatchSelection = (event) => {
      if (!(_batchSelectionRowId || _previousSelectionRange)) {
        return;
      }
      const checkboxElement = event.target.closest(
        "[data-row-selection-checkbox]"
      );
      if (checkboxElement === null) {
        _batchSelectionRowId = void 0;
        _previousSelectionRange = void 0;
        return;
      }
      if (table.getRowVirtualizationContainer()?.contains(checkboxElement) === false) {
        _batchSelectionRowId = void 0;
        _previousSelectionRange = void 0;
      }
    };
    table.getToggleAllLimitedRowsSelectedHandler = () => {
      return (e) => {
        table.toggleAllLimitedRowsSelected(
          e.target.checked
        );
      };
    };
    table.getToggleAllLimitedPageRowsSelectedHandler = () => {
      return (e) => {
        table.toggleAllLimitedPageRowsSelected(
          e.target.checked
        );
      };
    };
    table.getSelectionCount = () => table.getFilteredSelectedRowModel().flatRows.length;
    table.toggleAllPageRowsSelected = (value) => {
      table.setRowSelection((old) => {
        const resolvedValue = value ?? !table.getIsAllPageRowsSelected();
        const rowSelection = { ...old };
        table.getRowModel().flatRows.forEach((row) => {
          mutateRowIsSelected(
            rowSelection,
            row.id,
            resolvedValue,
            false,
            // we don't need to include children as they are already included in `flatRows`
            table
          );
        });
        return rowSelection;
      });
    };
    table.toggleAllRowsSelected = (value) => {
      table.setRowSelection((old) => {
        const resolvedValue = value ?? !table.getIsAllRowsSelected();
        const rowSelection = { ...old };
        table.getPrePaginationRowModel().flatRows.forEach((row) => {
          mutateRowIsSelected(
            rowSelection,
            row.id,
            resolvedValue,
            false,
            // we don't need to include children as they are already included in `flatRows`
            table
          );
        });
        return rowSelection;
      });
    };
    table.toggleAllLimitedPageRowsSelected = (value) => {
      const resolvedValue = value ?? !table.getIsAllPageRowsSelected();
      if (resolvedValue === false) {
        table.toggleAllPageRowsSelected(false);
        return;
      }
      table.setRowSelection((old) => {
        const rowSelection = { ...old };
        const limit = table.options.rowSelectionLimit ?? 0;
        const currentSelectionLength = table.getSelectionCount();
        if (currentSelectionLength < limit) {
          const rows = table.getRowModel().rows;
          let currentSelectionCount = currentSelectionLength;
          for (const row of rows) {
            if (!row.getCanSelect() || row.getIsSelected()) {
              continue;
            }
            rowSelection[row.id] = true;
            currentSelectionCount += 1;
            if (currentSelectionCount >= limit) {
              break;
            }
          }
        }
        return rowSelection;
      });
    };
    table.toggleAllLimitedRowsSelected = (value) => {
      const resolvedValue = value ?? !table.getIsAllPageRowsSelected();
      if (resolvedValue === false) {
        table.toggleAllRowsSelected(false);
        return;
      }
      table.setRowSelection((old) => {
        const rowSelection = { ...old };
        const limit = table.options.rowSelectionLimit ?? 0;
        const currentSelectionLength = table.getSelectionCount();
        if (currentSelectionLength < limit) {
          const rows = table.getPrePaginationRowModel().flatRows;
          let currentSelectionCount = currentSelectionLength;
          for (const row of rows) {
            if (!row.getCanSelect() || row.getIsSelected()) {
              continue;
            }
            rowSelection[row.id] = true;
            currentSelectionCount += 1;
            if (currentSelectionCount >= limit) {
              break;
            }
          }
        }
        return rowSelection;
      });
    };
    table.getIsRowSelectionLimitReached = (rowSelectionState) => {
      const rowSelection = rowSelectionState || table.getState().rowSelection;
      const rowSelectionLimit = table.options.rowSelectionLimit;
      const selectionCount = Object.keys(rowSelection).length;
      return rowSelectionLimit !== void 0 && rowSelectionLimit !== 0 && selectionCount >= rowSelectionLimit;
    };
    table.getIsAllPageRowsSelected = () => {
      if (table.getIsRowSelectionLimitReached()) {
        return true;
      }
      return table.getPaginationRowModel().flatRows.filter((row) => row.subRows.length === 0).every((row) => row.getIsSelected() || !row.getCanSelect());
    };
    table.getIsAllRowsSelected = () => {
      if (table.getIsRowSelectionLimitReached()) {
        return true;
      }
      return table.getPrePaginationRowModel().flatRows.filter((row) => row.subRows.length === 0).every((row) => row.getIsSelected() || !row.getCanSelect());
    };
    table.getIsSomePageRowsSelected = () => {
      return table.getPaginationRowModel().flatRows.filter((row) => row.subRows.length === 0).some((row) => row.getIsSelected());
    };
    table.getIsSomeRowsSelected = () => {
      return table.getPrePaginationRowModel().flatRows.filter((row) => row.subRows.length === 0).some((row) => row.getIsSelected());
    };
    table.getIsAllPageRowsSelectionDisabled = () => {
      return table.getPaginationRowModel().flatRows.every((row) => row.getShouldDisableRowSelection());
    };
    table.getIsAllRowsSelectionDisabled = () => {
      return table.getPrePaginationRowModel().flatRows.every((row) => row.getShouldDisableRowSelection());
    };
    (0, import_utils.assignRowMethods)(table, {
      getShouldDisableRowSelection() {
        const { enableRowSelection } = table.options;
        const isRowSelectionEnabled = typeof enableRowSelection === "function" ? enableRowSelection(this) : enableRowSelection ?? true;
        const hasSelectionLimit = table.options.rowSelectionLimit !== void 0;
        if (hasSelectionLimit && this.getIsSelected()) {
          return false;
        }
        return !isRowSelectionEnabled || table.getIsRowSelectionLimitReached();
      },
      getCanSelect() {
        const isParentRow = !!this.subRows?.length;
        if (isParentRow) {
          return false;
        }
        const { enableRowSelection } = table.options;
        return typeof enableRowSelection === "function" ? enableRowSelection(this) : enableRowSelection ?? true;
      },
      toggleSelected(value) {
        if (this.getShouldDisableRowSelection()) {
          return;
        }
        const isSelected = this.getIsSelected();
        table.setRowSelection((old) => {
          value = value ?? !isSelected;
          const selectedRowIds = { ...old };
          if (this.getCanSelect() && isSelected === value) {
            return old;
          }
          mutateRowIsSelected(selectedRowIds, this.id, value, true, table);
          return selectedRowIds;
        });
      },
      // Override TanStack's implementation to include selected, but disabled rows
      getIsSomeSelected() {
        const subRows = this.subRows;
        if (!subRows.length) {
          return this.getIsSelected();
        }
        return subRows.some(
          (subRow) => subRow.getIsSelected() || subRow.getIsSomeSelected()
        );
      },
      // Override TanStack's implementation to include selected, but disabled rows
      getIsAllSubRowsSelected() {
        const subRows = this.subRows;
        if (!subRows.length) {
          return this.getIsSelected();
        }
        for (const subRow of subRows) {
          if (!subRow.getIsAllSubRowsSelected()) {
            return false;
          }
        }
        return true;
      },
      /**
       * Handles row selection/deselection with support for batch selection using Shift+click. If Shift is not pressed,
       * or no previous selection exists, the row is simply selected/deselected and becomes the reference for a possible batch selection.
       * If a batch selection reference exists, and the user selects a row while holding Shift, all rows in between are selected/deselected.
       * If the user then selects another row with the Shift pressed, the previous batch selection is removed, and only the new one is selected.
       * The batch selection reference is reset, once the user clicks anywhere in the table outside the checkboxes.
       */
      batchSelect(value, shiftKey) {
        const selectSingleRow = () => {
          this.toggleSelected(value);
          _batchSelectionRowId = this.id;
          _previousSelectionRange = void 0;
        };
        if (_batchSelectionRowId === void 0 || !shiftKey) {
          selectSingleRow();
          return;
        }
        const flatRows = table.getRowOrderRowModel().flatRows;
        const firstSelectedIndex = flatRows.findIndex(
          (flatRow) => _batchSelectionRowId && flatRow.id === _batchSelectionRowId
        );
        const currentSelectedIndex = flatRows.findIndex(
          (flatRow) => flatRow.id === this.id
        );
        if (firstSelectedIndex === -1) {
          selectSingleRow();
          return;
        }
        const startIndex = Math.min(firstSelectedIndex, currentSelectedIndex);
        const endIndex = Math.max(firstSelectedIndex, currentSelectedIndex);
        const shouldUpdateChildren = (idx) => endIndex === idx;
        table.setRowSelection((old) => {
          const rowSelection = { ...old };
          const firstSelectedCheckedValue = flatRows[firstSelectedIndex].getIsSelected();
          if (_previousSelectionRange) {
            const start = _previousSelectionRange.start;
            const end = Math.min(
              _previousSelectionRange.end,
              flatRows.length - 1
            );
            for (let i = start; i <= end; i++) {
              mutateRowIsSelected(
                rowSelection,
                flatRows[i].id,
                false,
                true,
                table
              );
            }
          }
          for (let i = startIndex; i <= endIndex; i++) {
            mutateRowIsSelected(
              rowSelection,
              flatRows[i].id,
              firstSelectedCheckedValue,
              shouldUpdateChildren(i),
              table
            );
          }
          _previousSelectionRange = firstSelectedCheckedValue ? { start: startIndex, end: endIndex } : void 0;
          return rowSelection;
        });
      }
    });
  }
};
function useRowSelection(props, options) {
  const configSelectableRows = (0, import_useDataTableConfig.useDataTableConfigElement)("selectableRows");
  const selectableRows = props.selectableRows ?? configSelectableRows;
  const enableRowSelectionCallback = (0, import_react.useCallback)(
    (row) => {
      if (selectableRows !== void 0 && typeof selectableRows !== "boolean" && selectableRows.disableRowSelection) {
        return !selectableRows.disableRowSelection(row.original);
      }
      return true;
    },
    [selectableRows]
  );
  if (selectableRows !== void 0) {
    options.enableRowSelection = true;
    let rowSelectionMode = "page";
    if (selectableRows === true || selectableRows === false) {
      options.enableRowSelection = selectableRows;
    } else {
      const { selectAllBehavior, limit } = selectableRows;
      options.enableRowSelection = enableRowSelectionCallback;
      if (selectAllBehavior) {
        rowSelectionMode = selectAllBehavior;
      }
      options.rowSelectionLimit = limit !== void 0 && limit >= 0 ? limit : void 0;
    }
    options.enableMultiRowSelection = true;
    options.rowSelectionSelectAllMode = rowSelectionMode;
  } else {
    options.enableRowSelection = false;
  }
  const [selectedRows, setSelectedRows] = (0, import_useControllableState.useControllableState)({
    value: props.selectedRows,
    defaultValue: props.defaultSelectedRows,
    onChange: props.onRowSelectionChange
  });
  options.state = {
    ...options.state,
    rowSelection: selectedRows ?? {}
  };
  options.onRowSelectionChange = setSelectedRows;
}
