// @dynatrace/strato-components/esm/charts/core/components/legend/categorical/components/ChartLegendGrid.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import { ChartLegendItem } from "./ChartLegendItem/ChartLegendItem.js";
import { legendContent, legendGrid } from "./ChartVirtualizedLegendGrid.sty.js";
import { ORIENTATION } from "../../../../constants.js";
import { useChartLayoutRatio } from "../../../chart-layout/hooks/useChartLayoutRatio.js";
import {
  CATEGORICAL_LEGEND_CLASS,
  LEGEND_ACTION_BUTTON_WIDTH
} from "../constants.js";
import { useLegendGridSettings } from "../context/LegendGridSettingsContext.js";
const ChartLegendGrid = (props) => {
  const { items, orientation } = props;
  const { legendRef } = useChartLayoutRatio();
  const {
    maxRows,
    minItemWidth,
    canFitAllItemsWithoutScrollbar,
    maxItemWidth
  } = useLegendGridSettings();
  const nrOfColumns = Math.ceil(items.length / maxRows);
  const gridTemplateColumns = !canFitAllItemsWithoutScrollbar || orientation === ORIENTATION.Horizontal ? `repeat(auto-fit, minmax(${minItemWidth - LEGEND_ACTION_BUTTON_WIDTH}px, 1fr))` : `repeat(${nrOfColumns}, minmax(${minItemWidth}px, ${maxItemWidth}px))`;
  return /* @__PURE__ */ jsx("div", { className: legendContent({ layout: orientation }), children: /* @__PURE__ */ jsx(
    "div",
    {
      ref: legendRef,
      "data-testid": "chart-legend-grid-static",
      className: clsx(
        legendGrid({ layout: orientation }),
        CATEGORICAL_LEGEND_CLASS
      ),
      style: { gridTemplateColumns },
      children: items.map((item) => /* @__PURE__ */ jsx(
        ChartLegendItem,
        {
          item,
          numberOfItems: items.length
        },
        item.id
      ))
    }
  ) });
};
ChartLegendGrid["displayName"] = "ChartLegendGrid";
export {
  ChartLegendGrid
};
