// @dynatrace/strato-components/esm/charts/core/components/legend/categorical/components/ChartVirtualizedLegendGrid.js
import { jsx } from "react/jsx-runtime";
import { useVirtualizer } from "@tanstack/react-virtual";
import clsx from "clsx";
import { useCallback, useRef } from "react";
import { ChartLegendItem } from "./ChartLegendItem/ChartLegendItem.js";
import {
  legendContent,
  legendGrid,
  virtualizedGrid,
  virtualizedRow
} from "./ChartVirtualizedLegendGrid.sty.js";
import { useChartLayoutRatio } from "../../../chart-layout/hooks/useChartLayoutRatio.js";
import { useLegendSize } from "../../shared/context/LegendSizeContext.js";
import { CATEGORICAL_LEGEND_CLASS, ITEM_HEIGHT } from "../constants.js";
import { useLegendGridSettings } from "../context/LegendGridSettingsContext.js";
const ChartVirtualizedLegendGrid = (props) => {
  const { items, orientation } = props;
  const { legendRef } = useChartLayoutRatio();
  const { minItemWidth } = useLegendGridSettings();
  const { width: legendWidth } = useLegendSize();
  const seriesCount = items.length;
  const nrOfColumns = Math.max(Math.floor(legendWidth / minItemWidth), 1);
  const nrOfRows = Math.ceil(seriesCount / nrOfColumns);
  const parentRef = useRef(null);
  const virtualizer = useVirtualizer({
    count: nrOfRows,
    getScrollElement: () => parentRef.current,
    estimateSize: useCallback(() => ITEM_HEIGHT.value, []),
    overscan: 0
  });
  return /* @__PURE__ */ jsx(
    "div",
    {
      ref: parentRef,
      className: `${legendContent({
        layout: orientation
      })} chart-legend-grid-content`,
      children: /* @__PURE__ */ jsx(
        "div",
        {
          ref: legendRef,
          "data-testid": "chart-legend-grid-virtual",
          className: clsx(virtualizedGrid, CATEGORICAL_LEGEND_CLASS),
          style: {
            height: virtualizer.getTotalSize()
          },
          children: virtualizer.getVirtualItems().map((virtualRow) => {
            const rowStartIndex = virtualRow.index * nrOfColumns;
            const rowEndIndex = rowStartIndex + nrOfColumns;
            const gridTemplateColumns = `repeat(${nrOfColumns}, 1fr)`;
            return /* @__PURE__ */ jsx(
              "div",
              {
                className: virtualizedRow,
                style: {
                  height: `${virtualRow.size}px`,
                  transform: `translateY(${virtualRow.start}px)`
                },
                children: /* @__PURE__ */ jsx(
                  "div",
                  {
                    className: legendGrid({ layout: orientation }),
                    style: { gridTemplateColumns },
                    children: items.slice(rowStartIndex, rowEndIndex).map((item) => /* @__PURE__ */ jsx(
                      ChartLegendItem,
                      {
                        item,
                        numberOfItems: items.length
                      },
                      item.id
                    ))
                  }
                )
              },
              virtualRow.index
            );
          })
        }
      )
    }
  );
};
export {
  ChartVirtualizedLegendGrid
};
