// @dynatrace/strato-components/esm/tables/DataTable/features/ColumnTypes/ColumnTypes.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { memo } from "react";
import Colors from "@dynatrace/strato-design-tokens/colors";
import {
  adjustFractionDigits,
  format,
  formatLong,
  units
} from "@dynatrace-sdk/units";
import { logContentWrapper } from "./ColumnTypes.sty.js";
import { Sparkline } from "../../../../charts/sparkline/Sparkline.js";
import { Markdown } from "../../../../content/markdown/Markdown.js";
import {
  formatCurrencyDefault,
  formatDatetimeDefault
} from "../../../../core/utils/default-formatters.js";
import { devLog } from "../../../../core/utils/logging.js";
import { convertToNumberIfPureNumber } from "../../../../core/utils/string-utils.js";
import { LogContentHighlight } from "../../../utils/log-content-highlighting.js";
import { replaceUrlsWithLinks } from "../../../utils/replace-urls-with-links.js";
import { DATA_TABLE_DEFAULT_HORIZONTAL_ALIGNMENT } from "../../constants.js";
import {
  DataTableGanttChartCell,
  isGanttSegmentData
} from "../GanttChart/DataTableGanttChartCell.js";
import { DataTableGanttChartHeader } from "../GanttChart/DataTableGanttChartHeader.js";
import {
  isMultiMeterBarData,
  MeterBarChartRenderer
} from "../MeterBarChart/MeterBarChart.js";
function applyColumnTypes(columnDef, composedDef) {
  if (columnDef.cell !== void 0) {
    return {};
  }
  const columnType = columnDef.columnType;
  if (!columnType) {
    return {};
  }
  if (columnType === "date") {
    devLog(
      "DataTable: columnType 'date' is deprecated and will be removed in a future release. Use 'datetime' instead."
    );
  }
  let enableUserActions = true;
  if (columnType === "sparkline") {
    columnDef.disableSorting = true;
  }
  if (columnType === "meterbar") {
    columnDef.disableSorting ??= true;
    enableUserActions = { type: "chart" };
  }
  if (columnType === "markdown") {
    columnDef.disableSorting = true;
  }
  if (columnType === "gantt") {
    columnDef.disableSorting = true;
    enableUserActions = false;
    columnDef.header = () => /* @__PURE__ */ jsx(DataTableGanttChartHeader, { columnDef });
    composedDef.meta = {
      ...composedDef.meta,
      hasCustomCellRenderer: true,
      hasCustomHeaderRenderer: true
    };
  }
  if (columnType === "log-content") {
    composedDef.meta = {
      ...composedDef.meta,
      fontStyle: "code"
    };
  }
  return {
    alignment: columnDef.alignment || getAlignmentBasedOnType(columnType),
    cell: getCellTypeRenderer(columnType),
    enableSorting: !columnDef.disableSorting,
    header: columnDef.header,
    enableUserActions,
    meta: composedDef.meta
  };
}
function getAlignmentBasedOnType(columnType) {
  switch (columnType) {
    case "number":
    case "long":
    case "currency":
    case "bit":
    case "date":
    // DEPRECATION: [APPDEV-16044] BREAKING-CHANGE: Remove 'date' columnType.
    case "datetime":
      return "right";
    default:
      return DATA_TABLE_DEFAULT_HORIZONTAL_ALIGNMENT;
  }
}
function isDatatableGanttColumnConfig(config) {
  return config !== null && config !== void 0 && typeof config === "object" && "nameAccessor" in config;
}
function isDatatableSparklineColumnConfig(config) {
  return config !== null && config !== void 0 && typeof config === "object" && ("color" in config || "variant" in config || "gapPolicy" in config || "curve" in config || "showContextValues" in config || "xAxis" in config || "yAxis" in config);
}
function isDatatableMarkdownColumnConfig(config) {
  return config !== null && config !== void 0 && typeof config === "object" && "customComponentMappings" in config;
}
function isDatatableLogContentColumnConfig(config) {
  return config !== null && config !== void 0 && typeof config === "object" && "truncationLimit" in config;
}
function isDatatableTextColumnConfig(config) {
  return config !== null && config !== void 0 && typeof config === "object" && "detectLinks" in config;
}
const SparklineMemo = memo(Sparkline);
const MeterbarMemo = memo(MeterBarChartRenderer);
const MarkdownMemo = memo(Markdown);
const DatetimeMemo = memo((props) => {
  return /* @__PURE__ */ jsx(Fragment, { children: formatDatetimeDefault(
    new Date(convertToNumberIfPureNumber(props.value))
  ) });
});
const CELL_TYPE_RENDERERS = {
  number: (props) => {
    const cellValue = props.getValue();
    if (typeof cellValue === "string" || typeof cellValue === "number") {
      const valueAsNumber = Number.isNaN(+cellValue) || cellValue === null ? "" : convertToNumberIfPureNumber(cellValue);
      return typeof valueAsNumber === "number" ? adjustFractionDigits(valueAsNumber) : "";
    }
  },
  long: (props) => {
    const cellValue = props.getValue();
    if (cellValue === void 0 || cellValue === null || Number.isNaN(cellValue)) {
      return "";
    } else if (typeof cellValue === "number" || typeof cellValue === "bigint" || typeof cellValue === "string") {
      return formatLong(cellValue);
    }
    return `${cellValue}`;
  },
  currency: (props) => {
    const cellValue = props.getValue();
    if (typeof cellValue === "string" || typeof cellValue === "number") {
      const valueAsNumber = Number.isNaN(+cellValue) || cellValue === null ? "" : convertToNumberIfPureNumber(cellValue);
      return typeof valueAsNumber === "number" ? (
        // no currency (e.g. 'EUR', 'USD') provided → default format without currency symbol
        formatCurrencyDefault(valueAsNumber)
      ) : "";
    }
  },
  bit: (props) => {
    const cellValue = props.getValue();
    if (typeof cellValue === "string" || typeof cellValue === "number") {
      const valueAsNumber = Number.isNaN(+cellValue) ? "" : convertToNumberIfPureNumber(cellValue);
      return typeof valueAsNumber === "number" ? format(valueAsNumber, { input: units.data.bit }) : "";
    }
  },
  // DEPRECATION: [APPDEV-16044] BREAKING-CHANGE: Remove 'date' columnType.
  date: (props) => {
    const cellValue = props.getValue();
    if (cellValue === null) {
      return null;
    }
    return /* @__PURE__ */ jsx(DatetimeMemo, { value: cellValue });
  },
  datetime: (props) => {
    const cellValue = props.getValue();
    if (cellValue === null) {
      return null;
    }
    return /* @__PURE__ */ jsx(DatetimeMemo, { value: cellValue });
  },
  text: (props) => {
    const cellValue = props.getValue();
    let detectLinks = true;
    if (isDatatableTextColumnConfig(props.column.columnDef.meta?.config)) {
      detectLinks = !!props.column.columnDef.meta?.config.detectLinks;
    }
    if (typeof cellValue !== "string" || !detectLinks) {
      return cellValue === void 0 || cellValue === null || Number.isNaN(cellValue) ? "" : `${cellValue}`;
    }
    return replaceUrlsWithLinks(cellValue);
  },
  sparkline: (props) => {
    let config = {};
    if (props.column.columnDef.meta && isDatatableSparklineColumnConfig(props.column.columnDef.meta.config)) {
      config = props.column.columnDef.meta.config;
    }
    const rowDensity = props.table.getVisualVariantOptions().rowDensity;
    return /* @__PURE__ */ jsxs(
      SparklineMemo,
      {
        data: props.getValue(),
        color: config.color ?? Colors.Charts.Categorical.Color01.Default,
        variant: config.variant ?? "line",
        gapPolicy: config.gapPolicy ?? "connect",
        curve: config.curve ?? "linear",
        showContextValues: !!config?.showContextValues,
        labelsAlignment: "right",
        labelsFixedWidth: rowDensity === "condensed" ? 36 : 42,
        children: [
          config.xAxis && /* @__PURE__ */ jsx(Sparkline.XAxis, { ...config.xAxis }),
          config.yAxis && /* @__PURE__ */ jsx(Sparkline.YAxis, { ...config.yAxis })
        ]
      }
    );
  },
  meterbar: (props) => {
    const value = props.getValue();
    if (typeof value === "number" || isMultiMeterBarData(value)) {
      return /* @__PURE__ */ jsx(MeterbarMemo, { value, column: props.column });
    }
    return "";
  },
  gantt: (props) => {
    let config = { nameAccessor: "name" };
    if (props.column.columnDef.meta && isDatatableGanttColumnConfig(props.column.columnDef.meta.config)) {
      config = props.column.columnDef.meta.config;
    }
    const value = props.getValue();
    const parentData = props.row.getParentRow()?.getValue(props.column.id);
    const isGanttValue = isGanttSegmentData(value) || Array.isArray(value) && value.every(isGanttSegmentData);
    const isGanttParentData = parentData === void 0 || isGanttSegmentData(parentData) || Array.isArray(parentData) && parentData.every(isGanttSegmentData);
    const pageRows = props.table.getPaginationRowModel().rows;
    const isLastPageRow = props.row.id === pageRows.at(-1)?.id;
    if (isGanttValue && isGanttParentData) {
      return /* @__PURE__ */ jsx(
        DataTableGanttChartCell,
        {
          data: value,
          parentData,
          columnId: props.column.id,
          config,
          isLastPageRow
        }
      );
    }
    return "";
  },
  markdown: (props) => {
    const config = props.column.columnDef.meta?.config;
    const customComponentMappings = isDatatableMarkdownColumnConfig(config) ? config?.customComponentMappings : void 0;
    const value = props.getValue();
    if (value !== null && value !== void 0 && typeof value === "string") {
      return /* @__PURE__ */ jsx(MarkdownMemo, { customComponentMappings, children: value });
    }
    return "";
  },
  "log-content": (props) => {
    const cellValue = props.getValue();
    if (typeof cellValue === "string") {
      const isColumnLineWrap = props.column.getIsLineWrapped();
      const config = props.column.columnDef.meta?.config;
      const truncationLimit = isDatatableLogContentColumnConfig(config) ? config?.truncationLimit : void 0;
      return /* @__PURE__ */ jsx("div", { className: logContentWrapper({ linewrap: isColumnLineWrap }), children: /* @__PURE__ */ jsx(
        LogContentHighlight,
        {
          logContent: cellValue,
          lineWrap: isColumnLineWrap,
          truncationLimit
        }
      ) });
    }
    return "";
  }
};
function getCellTypeRenderer(columnType) {
  return CELL_TYPE_RENDERERS[columnType] ?? CELL_TYPE_RENDERERS.text;
}
export {
  applyColumnTypes,
  getAlignmentBasedOnType,
  isDatatableGanttColumnConfig,
  isDatatableLogContentColumnConfig
};
