// @dynatrace/strato-components/esm/charts/xy-chart/components/axes/XYChartRightAxisArea.js
import { jsx } from "react/jsx-runtime";
import { YTickWithTruncation } from "./YTickWithTruncation.js";
import { AxisRight } from "../../../core/components/axes/AxisRight.js";
import { DivAxisWrapper } from "../../../core/components/axes/components/DivAxisWrapper.js";
import {
  axisFontProps,
  EMOJI_REGEX
} from "../../../core/components/axes/constants.js";
import { getYTickMaxSize } from "../../../core/components/axes/utils/calculate-axis-size.js";
import { getTickLabelPropsForValueAxis } from "../../../core/components/axes/utils/get-tick-label-props.js";
import { isTickLabelSpaceAvailable } from "../../../core/components/axes/utils/is-tick-label-visible.js";
import { ChartGrid } from "../../../core/components/chart-grid/ChartGridRoot.js";
import { useAxesTicks } from "../../hooks/useAxesTicks.js";
import { useVisibleAxesConfig } from "../../hooks/useVisibleAxesConfig.js";
import { useVisibleAxesScales } from "../../hooks/useVisibleAxesScales.js";
import { isCategorical } from "../../types/guards.js";
import { AREAS } from "../../utils/grid-templates.js";
const tickLabelProps = getTickLabelPropsForValueAxis("right");
function XYChartRightAxisArea() {
  const { rightTicks } = useAxesTicks();
  const { right: rightConfig } = useVisibleAxesConfig();
  const { right: rightScale } = useVisibleAxesScales();
  const tickFormat = (_, index) => rightTicks.formattedTicks[index];
  const isCategoricalAxis = isCategorical(rightConfig?.type || "");
  const areCategoricalTicksNotFittable = rightScale && isCategoricalAxis && !isTickLabelSpaceAvailable(rightScale.step());
  const hideAxis = rightConfig === void 0 || areCategoricalTicksNotFittable || rightScale === void 0;
  const tickNameHasEmoji = rightTicks.ticks.some(
    (name) => EMOJI_REGEX.test(name)
  );
  return hideAxis ? null : /* @__PURE__ */ jsx(ChartGrid.GridArea, { gridArea: AREAS.rightAxis, children: ({ availableWidth, availableHeight }) => /* @__PURE__ */ jsx(
    DivAxisWrapper,
    {
      divAxisStyle: axisFontProps,
      ticks: rightTicks.formattedTicks,
      label: rightConfig.label,
      children: /* @__PURE__ */ jsx(
        AxisRight,
        {
          scale: rightScale,
          containerSize: availableHeight,
          width: availableWidth,
          tickValues: rightTicks.ticks,
          tickLabelProps,
          tickFormat,
          label: rightConfig.label,
          tickComponent: (renderProps) => /* @__PURE__ */ jsx(
            YTickWithTruncation,
            {
              position: "right",
              renderProps,
              maxSpaceForTruncation: getYTickMaxSize(
                availableWidth,
                rightConfig.label
              ),
              centerTick: isCategoricalAxis,
              tickNameHasEmoji
            }
          )
        }
      )
    }
  ) });
}
XYChartRightAxisArea["displayName"] = "XYChartRightAxisArea";
export {
  XYChartRightAxisArea
};
