// @dynatrace/strato-components/esm/charts/xy-chart/components/axes/XYChartBottomAxisArea.js
import { jsx, jsxs } from "react/jsx-runtime";
import { useGesture } from "@use-gesture/react";
import clsx from "clsx";
import { useRef } from "react";
import { axisContainerStyle } from "./AxisArea.sty.js";
import { computeCollisionAwareMaxSpace } from "./utils/compute-collision-aware-max-space.js";
import { isAxisPannable } from "./utils/is-axis-pannable.js";
import { XTickWithTruncation } from "./XTickWithTruncation.js";
import { XYChartPanningControls } from "./XYChartPanningControls.js";
import { AxisBottom } from "../../../core/components/axes/AxisBottom.js";
import { DivAxisWrapper } from "../../../core/components/axes/components/DivAxisWrapper.js";
import {
  axisFontProps,
  EMOJI_REGEX
} from "../../../core/components/axes/constants.js";
import { getXTickMaxSize } from "../../../core/components/axes/utils/calculate-axis-size.js";
import { getXAxisTickLabelProps } from "../../../core/components/axes/utils/get-tick-label-props.js";
import { isTickLabelSpaceAvailable } from "../../../core/components/axes/utils/is-tick-label-visible.js";
import { ChartGrid } from "../../../core/components/chart-grid/ChartGridRoot.js";
import { KEYBOARD_HANDLERS } from "../../../core/components/zoom/constants.js";
import { cursorStyle } from "../../../core/components/zoom/ZoomCursors.sty.js";
import { XY_CHART_BOTTOM_AXIS_LABEL } from "../../constants/defaults.js";
import { useAxesTicks } from "../../hooks/useAxesTicks.js";
import { useVisibleAxesConfig } from "../../hooks/useVisibleAxesConfig.js";
import { useVisibleAxesScales } from "../../hooks/useVisibleAxesScales.js";
import { useXYChartScales } from "../../hooks/useXYChartScales.js";
import { isCategorical } from "../../types/guards.js";
import { AREAS } from "../../utils/grid-templates.js";
import { useApplyPanDelta } from "../overlay/hooks/useApplyPanDelta.js";
function XYChartBottomAxisArea() {
  const axisRef = useRef(null);
  const { bottomTicks } = useAxesTicks();
  const { bottom: bottomConfig } = useVisibleAxesConfig();
  const { bottom: bottomScale } = useVisibleAxesScales();
  const scales = useXYChartScales();
  const canPan = isAxisPannable(bottomConfig, scales);
  const tickFormat = (_, index) => bottomTicks.formattedTicks[index];
  const isCategoricalAxis = bottomConfig?.type ? isCategorical(bottomConfig.type) : false;
  const isTickRotated = bottomConfig?.tickLabelLayout === "vertical";
  const tickLabelProps = getXAxisTickLabelProps(
    "bottom",
    isCategoricalAxis,
    isTickRotated,
    bottomScale,
    bottomConfig?.tickLabelLayout
  );
  const areCategoricalTicksNotFittable = bottomScale && isCategoricalAxis && !isTickLabelSpaceAvailable(bottomScale.step());
  const handlePan = useApplyPanDelta();
  useGesture(
    {
      onDrag: (e) => KEYBOARD_HANDLERS.includes(e.type) ? null : handlePan(-e.delta[0])
    },
    {
      target: axisRef,
      eventOptions: { passive: false },
      drag: { filterTaps: true, pointer: { buttons: [1, 4] } },
      enabled: canPan
    }
  );
  const hideAxis = bottomConfig === void 0 || areCategoricalTicksNotFittable || bottomScale === void 0;
  const tickNameHasEmoji = bottomTicks.ticks.some(
    (name) => EMOJI_REGEX.test(name)
  );
  return hideAxis ? null : /* @__PURE__ */ jsx(ChartGrid.GridArea, { gridArea: AREAS.bottomAxis, children: ({ availableWidth, availableHeight }) => {
    const maxSpaceForTruncation = computeCollisionAwareMaxSpace(
      getXTickMaxSize(
        availableHeight,
        availableWidth,
        isTickRotated,
        bottomScale,
        bottomConfig.label,
        bottomTicks.ticks.length
      ),
      bottomTicks.formattedTicks,
      bottomTicks.ticks,
      bottomScale,
      isCategoricalAxis,
      isTickRotated
    );
    return /* @__PURE__ */ jsxs(
      "div",
      {
        className: clsx(
          cursorStyle({ isPanEnabled: canPan }),
          axisContainerStyle
        ),
        ref: axisRef,
        children: [
          canPan && /* @__PURE__ */ jsx(
            XYChartPanningControls,
            {
              visibleAxis: bottomConfig,
              onClickPan: (delta) => handlePan(-delta)
            }
          ),
          /* @__PURE__ */ jsx(
            DivAxisWrapper,
            {
              divAxisStyle: axisFontProps,
              ticks: bottomTicks.formattedTicks,
              label: bottomConfig.label,
              isTickVertical: true,
              isTickRotated,
              children: /* @__PURE__ */ jsx(
                AxisBottom,
                {
                  role: "group",
                  "aria-label": XY_CHART_BOTTOM_AXIS_LABEL,
                  scale: bottomScale,
                  containerSize: availableWidth,
                  height: availableHeight,
                  tickValues: bottomTicks.ticks,
                  tickLabelProps,
                  tickFormat,
                  label: bottomConfig.label,
                  tickComponent: (renderProps) => /* @__PURE__ */ jsx(
                    XTickWithTruncation,
                    {
                      renderProps,
                      maxSpaceForTruncation,
                      centerTick: isCategoricalAxis,
                      isTickRotated,
                      position: "bottom",
                      tickNameHasEmoji
                    }
                  )
                }
              )
            }
          )
        ]
      }
    );
  } });
}
XYChartBottomAxisArea["displayName"] = "XYChartValueBottomAxisArea";
export {
  XYChartBottomAxisArea
};
