// @dynatrace/strato-components/esm/charts/xy-chart/components/axes/XYChartLeftAxisArea.js
import { jsx } from "react/jsx-runtime";
import { YTickWithTruncation } from "./YTickWithTruncation.js";
import { AxisLeft } from "../../../core/components/axes/AxisLeft.js";
import { DivAxisWrapper } from "../../../core/components/axes/components/DivAxisWrapper.js";
import {
  axisFontProps,
  EMOJI_REGEX
} from "../../../core/components/axes/constants.js";
import { getYTickMaxSize } from "../../../core/components/axes/utils/calculate-axis-size.js";
import { getTickLabelPropsForValueAxis } from "../../../core/components/axes/utils/get-tick-label-props.js";
import { isTickLabelSpaceAvailable } from "../../../core/components/axes/utils/is-tick-label-visible.js";
import { ChartGrid } from "../../../core/components/chart-grid/ChartGridRoot.js";
import { useAxesTicks } from "../../hooks/useAxesTicks.js";
import { useVisibleAxesConfig } from "../../hooks/useVisibleAxesConfig.js";
import { useVisibleAxesScales } from "../../hooks/useVisibleAxesScales.js";
import { isCategorical } from "../../types/guards.js";
import { AREAS } from "../../utils/grid-templates.js";
const tickLabelProps = getTickLabelPropsForValueAxis("left");
function XYChartLeftAxisArea() {
  const { leftTicks } = useAxesTicks();
  const { left: leftConfig } = useVisibleAxesConfig();
  const { left: leftScale } = useVisibleAxesScales();
  const tickFormat = (_, index) => leftTicks.formattedTicks[index];
  const isCategoricalAxis = leftConfig?.type ? isCategorical(leftConfig.type) : false;
  const areCategoricalTicksNotFittable = leftScale && isCategoricalAxis && !isTickLabelSpaceAvailable(leftScale.step());
  const hideAxis = leftConfig === void 0 || areCategoricalTicksNotFittable || leftScale === void 0;
  const tickNameHasEmoji = leftTicks.ticks.some(
    (name) => EMOJI_REGEX.test(name)
  );
  return hideAxis ? null : /* @__PURE__ */ jsx(ChartGrid.GridArea, { gridArea: AREAS.leftAxis, children: ({ availableWidth, availableHeight }) => /* @__PURE__ */ jsx(
    DivAxisWrapper,
    {
      divAxisStyle: axisFontProps,
      ticks: leftTicks.formattedTicks,
      label: leftConfig.label,
      children: /* @__PURE__ */ jsx(
        AxisLeft,
        {
          scale: leftScale,
          containerSize: availableHeight,
          width: availableWidth,
          tickValues: leftTicks.ticks,
          tickLabelProps,
          tickFormat,
          label: leftConfig.label,
          tickComponent: (renderProps) => /* @__PURE__ */ jsx(
            YTickWithTruncation,
            {
              position: "left",
              renderProps,
              maxSpaceForTruncation: getYTickMaxSize(
                availableWidth,
                leftConfig.label
              ),
              centerTick: isCategoricalAxis,
              tickNameHasEmoji
            }
          )
        }
      )
    }
  ) });
}
XYChartLeftAxisArea["displayName"] = "XYChartLeftAxisArea";
export {
  XYChartLeftAxisArea
};
