// @dynatrace/strato-components/charts/core/components/overlay-tooltip/components/global-actions/Action.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Action_exports = {};
__export(Action_exports, {
  Action: () => Action
});
module.exports = __toCommonJS(Action_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_actions_css = require("./actions.sty.js");
var import_Button = require("../../../../../../buttons/button/Button.js");
var import_useTooltipReducer = require("../../useTooltipReducer.js");
const Action = (props) => {
  const { label, icon, trailingIcon, onSelect, disabled } = props;
  const dispatch = (0, import_useTooltipReducer.useTooltipReducer)();
  const handleKeyDown = (event) => {
    if (event.key === "Enter" || event.key === " ") {
      event.preventDefault();
      event.stopPropagation();
      onSelect(event);
    }
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_Button.Button,
    {
      disabled,
      variant: "default",
      className: import_actions_css.actionButton,
      onClick: onSelect,
      onKeyDown: handleKeyDown,
      onPointerEnter: dispatch({
        type: "SET_ACTIVE_SERIES",
        payload: { activeSeries: "global-action" }
      }),
      width: "full",
      textAlign: "start",
      children: [
        icon && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: icon }),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Label, { children: label }),
        trailingIcon && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Suffix, { children: trailingIcon })
      ]
    }
  );
};
