// @dynatrace/strato-components/esm/charts/core/components/overlay-tooltip/components/DynamicTooltip.js
import { jsx } from "react/jsx-runtime";
import { useCallback, useEffect, useMemo } from "react";
import { TooltipPortal } from "./TooltipPortal.js";
import { useThrottledCallbackOnResize } from "../../../hooks/useThrottledCallbackOnResize.js";
import { useThrottledCallbackOnScroll } from "../../../hooks/useThrottledCallbackOnScroll.js";
import { throttleRaf } from "../../../utils/throttleRaf.js";
import { useOverlay } from "../contexts/OverlayContext.js";
import { useTooltip } from "../contexts/TooltipContext.js";
import { useTooltipReducer } from "../useTooltipReducer.js";
const DynamicTooltip = (props) => {
  const { absoluteX, absoluteY, pinned } = props;
  const overlay = useOverlay();
  const dispatch = useTooltipReducer();
  const tooltip = useTooltip();
  const onEscapeKeyPressedHandler = useCallback((event) => {
    if (event.key === "Escape") {
      dispatch({ type: "RESET_TOOLTIP" })();
    }
  }, []);
  useEffect(() => {
    window.addEventListener("keyup", onEscapeKeyPressedHandler);
    return () => {
      window.removeEventListener("keyup", onEscapeKeyPressedHandler);
    };
  }, []);
  useThrottledCallbackOnScroll(false, () => {
    if (!overlay.overlayRef.current) {
      return;
    }
    if (tooltip.getState().pinned) {
      dispatch({
        type: "UPDATE_POSITION",
        payload: { type: "non-resize", scroll: true }
      })();
      return;
    }
    dispatch({ type: "SET_OUT_BOUNDS" })();
  });
  useThrottledCallbackOnResize(
    overlay.overlayRef,
    dispatch({ type: "UPDATE_POSITION", payload: { type: "resize" } }),
    !pinned
  );
  const pointerMoveCallback = useMemo(
    () => throttleRaf((event) => {
      if (event.pressure && tooltip.getState().pinned) {
        if (!overlay.overlayRef.current || overlay.overlayRef.current.contains(event.target)) {
          return;
        }
        dispatch({
          type: "UPDATE_POSITION",
          payload: { type: "non-resize" }
        })();
        return;
      }
    }),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    []
  );
  const pointerUpCallback = useMemo(
    () => throttleRaf((event) => {
      if (tooltip.getState().pinned) {
        if (!overlay.overlayRef.current || overlay.overlayRef.current.contains(event.target)) {
          return;
        }
        dispatch({
          type: "UPDATE_POSITION",
          payload: { type: "non-resize" }
        })();
        return;
      }
    }),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    []
  );
  useEffect(() => {
    if (!pinned || !overlay.overlayRef.current) {
      return;
    }
    window.addEventListener("pointermove", pointerMoveCallback);
    window.addEventListener("pointerup", pointerUpCallback);
    return () => {
      window.removeEventListener("pointermove", pointerMoveCallback);
      window.removeEventListener("pointerup", pointerUpCallback);
    };
  }, [pinned]);
  return /* @__PURE__ */ jsx(
    TooltipPortal,
    {
      isInteractive: pinned,
      absoluteX,
      absoluteY,
      onMouseLeave: dispatch({ type: "TOOLTIP_MOUSE_LEAVE" }),
      onScrollCapture: () => dispatch({ type: "TOOLTIP_MOUSE_LEAVE" }),
      children: props.children
    }
  );
};
export {
  DynamicTooltip
};
