// @dynatrace/strato-components/esm/charts/core/components/overlay-tooltip/components/TooltipPortal.js
import { jsx } from "react/jsx-runtime";
import { useState } from "react";
import { createPortal } from "react-dom";
import { OverlayStackingProvider } from "../../../../../core/providers/OverlayStackingProvider.js";
import { useDynamicTooltip } from "../hooks/useDynamicTooltip.js";
import { TooltipContainerDimensionProvider } from "../providers/tooltip-container-dimensions.Provider.js";
import { wrapper } from "../tooltip.sty.js";
const TooltipPortal = (props) => {
  const { absoluteX, absoluteY, isInteractive, onMouseLeave, onScrollCapture } = props;
  const [tooltipWrapper, setTooltipWrapper] = useState(
    null
  );
  const dynamicTooltip = useDynamicTooltip({
    template: tooltipWrapper,
    x: absoluteX,
    y: absoluteY
  });
  return createPortal(
    /* @__PURE__ */ jsx(
      "div",
      {
        role: "tooltip",
        "aria-label": "OverlayTooltip",
        ref: setTooltipWrapper,
        className: wrapper({ isInteractive }),
        onMouseLeave,
        onScrollCapture,
        ...dynamicTooltip,
        children: /* @__PURE__ */ jsx(OverlayStackingProvider, { children: /* @__PURE__ */ jsx(TooltipContainerDimensionProvider, { container: tooltipWrapper, children: props.children }) })
      }
    ),
    document.body
  );
};
export {
  TooltipPortal
};
