// @dynatrace/strato-components/esm/forms/calendar/components/CalendarItem.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import {
  forwardRef,
  useMemo
} from "react";
import * as itemStyles from "./CalendarItem.sty.js";
import { Button } from "../../../buttons/button/Button.js";
import { useFormFieldContext } from "../../forms-core/hooks/useFormFieldContext.js";
import { useCalendarContext } from "../hooks/useCalendarContext.js";
const CalendarItem = /* @__PURE__ */ forwardRef(
  ({
    children,
    isCurrent = false,
    highlightColor = "neutral",
    onClick,
    isSelected = false,
    isDisabled = false,
    isCurrentMonthDay = true,
    view,
    ariaLabel,
    ...remainingProps
  }, ref) => {
    const { disabled, ariaDisabled } = useFormFieldContext();
    const { readOnly } = useCalendarContext();
    const variant = useMemo(() => {
      if (view === "day") {
        return "default";
      }
      if (isSelected) {
        return "accent";
      }
      if (isCurrent) {
        return "emphasized";
      }
      return "default";
    }, [view, isSelected, isCurrent]);
    const currentStyle = view !== "day" && isCurrent ? itemStyles.currentIndicator : "";
    const selectionHandler = !isDisabled && !disabled && !readOnly ? {
      onClick,
      onKeyDown: (event) => {
        const { code } = event;
        if (code === "Space" || code === "Enter") {
          event.preventDefault();
          onClick?.();
        }
      }
    } : {};
    return /* @__PURE__ */ jsx(
      Button,
      {
        as: "div",
        ref,
        "aria-label": ariaLabel,
        ...remainingProps,
        className: clsx(itemStyles.calendarButton, currentStyle, {
          [itemStyles.outOfScopeDay]: !isCurrentMonthDay && highlightColor === "neutral" && !(disabled || isDisabled || ariaDisabled)
        }),
        color: highlightColor,
        disabled: disabled || isDisabled,
        "aria-disabled": ariaDisabled,
        variant,
        ...selectionHandler,
        children
      }
    );
  }
);
CalendarItem.displayName = "CalendarBodyDateButton";
export {
  CalendarItem
};
