// @dynatrace/strato-components/esm/forms/calendar/components/Days.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import { useContext, useMemo } from "react";
import { Day } from "./Day.js";
import * as styles from "../Calendar.sty.js";
import { DayViewContext } from "../contexts/DayViewContext.js";
import { useCalendarContext } from "../hooks/useCalendarContext.js";
import {
  NUMBER_OF_WEEK_ROWS,
  NUMBER_OF_WEEKDAYS
} from "../types/calendarConfiguration.js";
import { getDays } from "../utils/helpers.js";
function Days({ weekStartsOn }) {
  const { state } = useCalendarContext();
  const { setHoveredDate } = useContext(DayViewContext);
  const days = useMemo(
    () => getDays(state.activeDate, weekStartsOn),
    [state.activeDate, weekStartsOn]
  );
  return /* @__PURE__ */ jsx(
    "tbody",
    {
      onMouseLeave: () => {
        setHoveredDate?.(null);
      },
      children: new Array(NUMBER_OF_WEEK_ROWS).fill(0).map((_, index) => /* @__PURE__ */ jsx(
        "tr",
        {
          className: clsx([styles.row, styles.dayRow]),
          children: days.slice(
            index * NUMBER_OF_WEEKDAYS,
            index * NUMBER_OF_WEEKDAYS + NUMBER_OF_WEEKDAYS
          ).map((day, index2) => /* @__PURE__ */ jsx(
            Day,
            {
              date: day.date,
              position: index2 + 1,
              isCurrentMonthDay: day.isCurrentMonthDay
            },
            day.date.toISOString()
          ))
        },
        `${days[index].date}-row-${index}`
      ))
    }
  );
}
export {
  Days
};
