// @dynatrace/strato-components/esm/forms/calendar/components/Month.js
import { jsx } from "react/jsx-runtime";
import { useRef } from "react";
import { formatDate } from "@dynatrace-sdk/units";
import { field as monthFieldStyle } from "../Calendar.sty.js";
import { CalendarItem } from "./CalendarItem.js";
import { useFormFieldContext } from "../../forms-core/hooks/useFormFieldContext.js";
import { useFormFieldMessagesContext } from "../../forms-core/hooks/useFormFieldMessagesContext.js";
import { useMonth } from "../hooks/useMonth.js";
const Month = ({ date }) => {
  const itemRef = useRef(null);
  const { ariaLabel, tabIndex, isSelected, isCurrent, onClick, isDisabled } = useMonth({
    date,
    itemRef
  });
  const { disabled, ariaDisabled } = useFormFieldContext();
  const {
    state: { showError }
  } = useFormFieldMessagesContext();
  const highlightColor = isSelected ? "primary" : "neutral";
  const validatedHighlightColor = showError && highlightColor === "primary" ? "critical" : highlightColor;
  return /* @__PURE__ */ jsx(
    "td",
    {
      className: monthFieldStyle({
        color: validatedHighlightColor,
        isDisabled: isDisabled || disabled || ariaDisabled
      }),
      role: "presentation",
      children: /* @__PURE__ */ jsx(
        CalendarItem,
        {
          ref: itemRef,
          className: "calendar-button calendar-body-button",
          isCurrent,
          isSelected,
          isDisabled: isDisabled || disabled,
          highlightColor: validatedHighlightColor,
          tabIndex,
          role: "gridcell",
          "aria-selected": !!isSelected,
          ariaLabel,
          view: "month",
          onClick,
          "data-dt-name": "SelectMonth",
          "data-dt-properties": `value:${formatDate(date, {
            month: "numeric"
          })}`,
          children: formatDate(date, {
            month: "short"
          })
        }
      )
    }
  );
};
Month["displayName"] = "Month";
export {
  Month
};
