// @dynatrace/strato-components/esm/forms/calendar/components/Weekdays.js
import { jsx } from "react/jsx-runtime";
import { TZDateMini } from "@date-fns/tz";
import clsx from "clsx";
import { startOfWeek, addDays } from "date-fns";
import { formatDate } from "@dynatrace-sdk/units";
import { Text } from "../../../typography/text/Text.js";
import { eachDayOfInterval } from "../../forms-core/time/helpers.js";
import * as styles from "../Calendar.sty.js";
function Weekdays({
  weekStartsOn,
  configuredTimezone
}) {
  const today = TZDateMini.tz(configuredTimezone);
  const intervalStart = startOfWeek(today, {
    weekStartsOn
  });
  const days = eachDayOfInterval({
    start: intervalStart,
    end: addDays(intervalStart, 6)
  });
  return /* @__PURE__ */ jsx("thead", { children: /* @__PURE__ */ jsx("tr", { className: clsx([styles.row, styles.dayRow, styles.weekdays]), children: days.map((weekday, index) => /* @__PURE__ */ jsx(
    "th",
    {
      className: "weekday",
      abbr: formatDate(weekday, {
        weekday: "long",
        timeZone: configuredTimezone
      }),
      scope: "col",
      children: /* @__PURE__ */ jsx(Text, { textStyle: "small", as: "p", children: formatDate(weekday, {
        weekday: "short",
        timeZone: configuredTimezone
      }) })
    },
    `${weekday}-${index}`
  )) }) });
}
export {
  Weekdays
};
