// @dynatrace/strato-components/esm/forms/calendar/components/Year.js
import { jsx } from "react/jsx-runtime";
import { useRef } from "react";
import { formatDate } from "@dynatrace-sdk/units";
import { field as yearFieldStyle } from "../Calendar.sty.js";
import { CalendarItem } from "./CalendarItem.js";
import { useFormFieldContext } from "../../forms-core/hooks/useFormFieldContext.js";
import { useFormFieldMessagesContext } from "../../forms-core/hooks/useFormFieldMessagesContext.js";
import { useYear } from "../hooks/useYear.js";
const Year = ({ date }) => {
  const itemRef = useRef(null);
  const { ariaLabel, tabIndex, isCurrent, isSelected, onClick, isDisabled } = useYear({
    date,
    itemRef
  });
  const { disabled, ariaDisabled } = useFormFieldContext();
  const {
    state: { showError }
  } = useFormFieldMessagesContext();
  const highlightColor = isSelected ? "primary" : "neutral";
  const validatedHighlightColor = showError && highlightColor === "primary" ? "critical" : highlightColor;
  const yearValue = formatDate(date, { year: "numeric" });
  return /* @__PURE__ */ jsx(
    "td",
    {
      role: "presentation",
      className: yearFieldStyle({
        color: validatedHighlightColor,
        isDisabled: isDisabled || disabled || ariaDisabled
      }),
      children: /* @__PURE__ */ jsx(
        CalendarItem,
        {
          ref: itemRef,
          className: "calendar-button calendar-body-button",
          isCurrent,
          isSelected,
          isDisabled: isDisabled || disabled,
          highlightColor: validatedHighlightColor,
          tabIndex,
          role: "gridcell",
          "aria-selected": !!isSelected,
          ariaLabel,
          onClick,
          view: "year",
          "data-dt-name": "SelectYear",
          "data-dt-properties": `value:${yearValue}`,
          children: yearValue
        }
      )
    }
  );
};
Year["displayName"] = "Year";
export {
  Year
};
