// @dynatrace/strato-components/charts/gantt-chart/GanttChartLine.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var GanttChartLine_exports = {};
__export(GanttChartLine_exports, {
  GanttChartLine: () => GanttChartLine
});
module.exports = __toCommonJS(GanttChartLine_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_d3_scale = require("d3-scale");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_GanttChartLineRenderer = require("./components/GanttChartLineRenderer.js");
var import_GanttChartTooltip = require("./components/tooltip/GanttChartTooltip.js");
var import_defaults = require("./constants/defaults.js");
var import_GanttSegmentsContext = require("./context/GanttSegmentsContext.js");
var styles = __toESM(require("./GanttChartLine.sty.js"));
var import_use_set_tooltip_state = require("./hooks/use-set-tooltip-state.js");
var import_messages = require("./messages.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_useObserveContainerSize = require("../core/hooks/useObserveContainerSize.js");
const GanttChartLine = (0, import_react.forwardRef)((props, forwardedRef) => {
  const intl = (0, import_react_intl.useIntl)();
  const [mouseInside, setMouseInside] = (0, import_react.useState)(false);
  const isTooltipVisible = (0, import_use_set_tooltip_state.useTooltipVisibility)();
  const isTooltipPinned = (0, import_use_set_tooltip_state.useTooltipPinnedStatus)();
  const { size, segmentsData, parentSegmentsData, config, loading } = props;
  const [{ width }, containerRef] = (0, import_useObserveContainerSize.useObserveContainerSize)();
  const ref = (0, import_useMergeRefs.useMergeRefs)([forwardedRef, containerRef]);
  const height = import_defaults.GANTT_LINE_SIZE_HEIGHT[size ?? import_defaults.DEFAULT_GANTT_LINE_SIZE];
  const segmentScale = (0, import_d3_scale.scaleLinear)().domain([props.min, props.max]).range([0, width]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      className: styles.wrapper,
      onMouseEnter: () => !isTooltipPinned && setMouseInside(true),
      onMouseLeave: () => !isTooltipPinned && setMouseInside(false),
      role: "img",
      "aria-label": import_defaults.GANTT_CHART_V2_ARIA_LABEL,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
        import_GanttSegmentsContext.GanttSegmentsContext.Provider,
        {
          value: {
            segmentsData,
            parentSegmentsData,
            segmentScale,
            config,
            hovering: mouseInside
          },
          children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
              import_GanttChartLineRenderer.GanttChartLineRenderer,
              {
                width,
                height,
                loading,
                ref,
                showBackground: config.showBackground,
                children: [
                  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_GanttChartLineRenderer.GanttChartLineRenderer.EmptyState, { children: intl.formatMessage(import_messages.ganttChartMessages.emptyState) }),
                  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_GanttChartLineRenderer.GanttChartLineRenderer.ErrorState, { children: (errorMessage) => /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("span", { children: [
                    intl.formatMessage(import_messages.ganttChartMessages.errorState),
                    ":",
                    " ",
                    errorMessage
                  ] }) })
                ]
              }
            ),
            isTooltipVisible && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_GanttChartTooltip.GanttChartTooltip, {})
          ]
        }
      )
    }
  );
});
