// @dynatrace/strato-components/content/terminology-overlay/TerminologyOverlay.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TerminologyOverlay_exports = {};
__export(TerminologyOverlay_exports, {
  TerminologyOverlay: () => TerminologyOverlay
});
module.exports = __toCommonJS(TerminologyOverlay_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_Description = require("./Description.js");
var import_Footer = require("./Footer.js");
var import_terminology_overlay_constants = require("./terminology-overlay-constants.js");
var import_TerminologyOverlayContext = require("./TerminologyOverlayContext.js");
var import_Trigger = require("./Trigger.js");
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useId = require("../../core/hooks/useId.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_useSafeForwardProps = require("../../core/hooks/useSafeForwardProps.js");
var import_useSlots = require("../../core/hooks/useSlots.js");
var import_logging = require("../../core/utils/logging.js");
var import_Flex = require("../../layouts/flex/Flex.js");
var import_useBreakpoint = require("../../layouts/hooks/useBreakpoint.js");
var import_useOverlayWithTrigger = require("../../overlays/hooks/useOverlayWithTrigger.js");
var import_Overlay = require("../../overlays/overlay/Overlay.js");
const TerminologyOverlayRoot = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    children,
    defaultOpen,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const ariaLabelingProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(restProps);
  const forwardProps = (0, import_useSafeForwardProps.useSafeForwardProps)(restProps, ariaLabelingProps);
  const smallerScreens = (0, import_useBreakpoint.useBreakpoint)("(max-width: 480px)");
  const overlayContentId = (0, import_useId.useId)();
  const {
    targetRef,
    triggerRef,
    overlayRef,
    overlayState,
    overlayProps,
    overlayTriggerProps
  } = (0, import_useOverlayWithTrigger.useOverlayWithTrigger)({
    placement: "bottom-start",
    placementFallbacks: ["bottom-end", "top-start", "top-end"],
    offset: 4,
    crossOffset: -16,
    widthStrategy: smallerScreens ? { min: 200 } : { min: 350 },
    defaultOpen,
    shouldCloseOnBlur: true
  });
  const rootRef = (0, import_useMergeRefs.useMergeRefs)([forwardedRef, triggerRef, targetRef]);
  const [slots] = (0, import_useSlots.useSlots)(children, {
    trigger: import_Trigger.Trigger,
    description: import_Description.Description,
    footer: { match: import_Footer.Footer, multiple: true }
  });
  if (slots.trigger === void 0) {
    (0, import_logging.devLog)("Trigger element has not been defined!");
  }
  const { children: triggerChildren, ...triggerSlotProps } = slots.trigger?.props ?? {};
  const triggerProps = slots.trigger?.props ?? {};
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_TerminologyOverlayContext.TerminologyOverlayContext.Provider, { value: overlayTriggerProps, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Trigger.Trigger,
      {
        ref: rootRef,
        className: triggerProps.className,
        style: triggerProps.style,
        "aria-describedby": overlayState.isOpen ? overlayContentId : void 0,
        "data-testid": triggerProps["data-testid"],
        "data-dtrum-mask": triggerProps["data-dtrum-mask"] ?? dataDtrumMask,
        "data-dtrum-allow": triggerProps["data-dtrum-allow"] ?? dataDtrumAllow,
        ...triggerSlotProps,
        children: triggerChildren
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Overlay.Overlay,
      {
        ref: overlayRef,
        isOpen: overlayState.isOpen,
        overlayProps,
        className: consumerClassName,
        style: consumerStyle,
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        ...ariaLabelingProps,
        ...behavioralTrackingProps,
        ...forwardProps,
        "data-dt-component": import_terminology_overlay_constants.COMPONENT_NAME,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
          import_Flex.Flex,
          {
            id: overlayContentId,
            paddingY: 12,
            paddingX: 16,
            flexDirection: "column",
            children: [
              slots.description,
              slots.footer
            ]
          }
        )
      }
    )
  ] });
});
TerminologyOverlayRoot.displayName = import_terminology_overlay_constants.COMPONENT_NAME;
const TerminologyOverlay = Object.assign(TerminologyOverlayRoot, {
  Description: import_Description.Description,
  Trigger: import_Trigger.Trigger,
  Footer: import_Footer.Footer
});
