// @dynatrace/strato-components/content/terminology-overlay/Trigger.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Trigger_exports = {};
__export(Trigger_exports, {
  Trigger: () => Trigger
});
module.exports = __toCommonJS(Trigger_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_terminology_overlay_constants = require("./terminology-overlay-constants.js");
var import_TerminologyOverlay_css = require("./TerminologyOverlay.sty.js");
var import_TerminologyOverlayContext = require("./TerminologyOverlayContext.js");
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useFocusRing = require("../../core/styles/useFocusRing.js");
var import_merge_props = require("../../core/utils/merge-props.js");
const Trigger = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    children,
    id: propId,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const {
    "data-dt-name": dataDtName = "Trigger",
    ...remainingBehavioralTrackingProps
  } = behavioralTrackingProps;
  const ariaLabelingProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(restProps);
  const { focusProps, focusClassName, isFocusVisible } = (0, import_useFocusRing.useFocusRing)({
    isMinimal: true
  });
  const overlayTriggerProps = (0, import_react.useContext)(import_TerminologyOverlayContext.TerminologyOverlayContext);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "button",
    {
      className: (0, import_clsx.default)(
        focusClassName,
        import_TerminologyOverlay_css.trigger,
        {
          "strato-focus-visible": isFocusVisible
        },
        consumerClassName
      ),
      style: {
        textDecoration: isFocusVisible ? "none" : void 0,
        ...consumerStyle
      },
      ref: forwardedRef,
      ...(0, import_merge_props.mergeProps)(
        overlayTriggerProps,
        focusProps,
        ariaLabelingProps,
        remainingBehavioralTrackingProps
      ),
      "aria-haspopup": "dialog",
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "data-dt-name": dataDtName,
      "data-dt-component": import_terminology_overlay_constants.COMPONENT_NAME,
      children
    }
  );
});
Trigger.displayName = "TerminologyOverlay.Trigger";
