// @dynatrace/strato-components/navigation/breadcrumbs/Breadcrumbs.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Breadcrumbs_exports = {};
__export(Breadcrumbs_exports, {
  Breadcrumbs: () => Breadcrumbs
});
module.exports = __toCommonJS(Breadcrumbs_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
var import_Breadcrumbs_css = require("./Breadcrumbs.sty.js");
var import_Item = require("./Item.js");
var import_Button = require("../../buttons/button/Button.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useContainerSize = require("../../core/hooks/useContainerSize.js");
var import_useMeasureChildren = require("../../core/hooks/useMeasureChildren.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_Menu = require("../menu/Menu.js");
const COMPONENT_NAME = "Breadcrumbs";
const BreadcrumbsRoot = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    children,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    ...remainingProps
  } = props;
  const _containerRef = (0, import_react.useRef)(null);
  const { containerRef, measurements, childProps } = (0, import_useMeasureChildren.useMeasureChildren)();
  const { width: containerWidth } = (0, import_useContainerSize.useContainerSize)(_containerRef);
  const rootRef = (0, import_useMergeRefs.useMergeRefs)([forwardedRef, _containerRef, containerRef]);
  const overflowingBreadcrumbs = (0, import_react.useRef)(/* @__PURE__ */ new Map());
  const intl = (0, import_react_intl.useIntl)();
  const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const slots = (0, import_react.useMemo)(() => {
    const slots2 = { items: [], menuItems: [] };
    const childrenArray = import_react.Children.toArray(children);
    let childWidths = 0;
    for (const value of Object.values(measurements)) {
      childWidths += value?.width || 0;
    }
    if (Object.keys(measurements).length === childProps.length) {
      const childKeys = childProps.map((props2) => props2["data-id"]);
      let childIndex = overflowingBreadcrumbs.current.size ? 2 : 1;
      overflowingBreadcrumbs.current.clear();
      while (childWidths > containerWidth && childIndex < childProps.length) {
        if (childIndex !== childProps.length - 1) {
          const key = childKeys[childIndex];
          if (key) {
            overflowingBreadcrumbs.current.set(key, true);
            childWidths -= measurements[key]?.width || 0;
          }
        } else {
          const key = childKeys[0];
          if (key) {
            overflowingBreadcrumbs.current.set(key, true);
            childWidths -= measurements[key]?.width || 0;
          }
        }
        childIndex++;
      }
    }
    for (const [index, child] of childrenArray.entries()) {
      if ((0, import_Item.isBreadcrumbItem)(child)) {
        const uniqueKey = `nav-items-${index}`;
        const adjustedIndex = overflowingBreadcrumbs.current.size > 0 && index > 0 ? index + 1 : index;
        const isLast = index === childrenArray.length - 1;
        slots2.items.push(
          (0, import_react.cloneElement)(child, {
            ...child.props,
            key: uniqueKey,
            className: (0, import_clsx.default)(
              {
                [`${import_Breadcrumbs_css.breadcrumbsItemInvisibleCSS}`]: overflowingBreadcrumbs.current.get(
                  childProps[adjustedIndex]?.["data-id"] || ""
                )
              },
              child.props.className
            ),
            isLast,
            isLastVisible: isLast && overflowingBreadcrumbs.current.size === childrenArray.length - 1,
            isCollapsed: false,
            ...childProps[adjustedIndex]
          })
        );
        if (overflowingBreadcrumbs.current.get(
          childProps[adjustedIndex]?.["data-id"] || ""
        )) {
          slots2.menuItems.push(
            (0, import_react.cloneElement)(/* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Link, {}), {
              ...child.props,
              key: uniqueKey,
              isCollapsed: true,
              ...childProps[adjustedIndex]
            })
          );
        }
      }
    }
    return slots2;
  }, [children, measurements, containerWidth, childProps]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "nav",
    {
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "aria-label": intl.formatMessage({
        defaultMessage: "Breadcrumb",
        id: "x2bmZRg98b0KodlW"
      }),
      className: consumerClassName,
      style: consumerStyle,
      ...restProps,
      ...behavioralTrackingProps,
      "data-dt-component": COMPONENT_NAME,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("ol", { ref: rootRef, className: import_Breadcrumbs_css.breadcrumbsContainerCSS, children: [
        slots.items[0],
        overflowingBreadcrumbs.current.size > 0 && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Menu.Menu, { children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Trigger, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Item.Item, { as: import_Button.Button, ...childProps[1], menuTrigger: true, size: "condensed", children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { style: { color: import_colors.default.Text.Neutral.Default }, children: "\u2026" }) }) }),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Content, { children: slots.menuItems })
        ] }),
        slots.items.slice(1)
      ] })
    }
  );
});
BreadcrumbsRoot.displayName = "Breadcrumbs";
const Breadcrumbs = Object.assign(BreadcrumbsRoot, {
  Item: import_Item.Item
});
