// @dynatrace/strato-components/navigation/breadcrumbs/Item.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Item_exports = {};
__export(Item_exports, {
  Item: () => Item,
  isBreadcrumbItem: () => isBreadcrumbItem
});
module.exports = __toCommonJS(Item_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_is = require("react-is");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_Breadcrumbs_css = require("./Breadcrumbs.sty.js");
var import_focusRingStatic_css = require("../../core/styles/focusRingStatic.sty.js");
var import_Link = require("../../typography/link/Link.js");
function isBreadcrumbItem(node) {
  return node !== void 0 && node !== null && (0, import_react_is.isElement)(node) && node.type === Item;
}
const Item = /* @__PURE__ */ (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const {
      children,
      disabled = false,
      href,
      isLast,
      isLastVisible,
      isCollapsed,
      className: consumerClassName,
      style: consumerStyle,
      "data-id": dataId,
      "data-testid": dataTestId,
      as = "a",
      ...remainingProps
    } = props;
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      "li",
      {
        className: (0, import_clsx.default)(
          (0, import_Breadcrumbs_css.breadcrumbsListItemCSS)({ isLastVisible }),
          consumerClassName
        ),
        style: consumerStyle,
        "data-id": dataId,
        "data-testid": dataTestId,
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_Link.Link,
            {
              className: (0, import_clsx.default)(
                (0, import_focusRingStatic_css.focusRingStatic)({
                  minimal: true,
                  color: remainingProps.menuTrigger ? "neutral" : "primary",
                  disabled
                }),
                (0, import_Breadcrumbs_css.breadcrumbsItemCSS)({
                  disabled,
                  isLastItem: isLast
                })
              ),
              ref: forwardedRef,
              as,
              href,
              "aria-disabled": disabled,
              "aria-current": isLast && "page",
              ...remainingProps,
              tabIndex: !disabled ? 0 : -1,
              children
            }
          ),
          !isCollapsed && !isLast && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ChevronRightSmallIcon, { className: import_Breadcrumbs_css.breadcrumbsSeperatorIconCSS })
        ]
      }
    );
  }
);
Item.displayName = "Breadcrumb.Item";
