// @dynatrace/strato-components/forms/switch/Switch.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Switch_exports = {};
__export(Switch_exports, {
  Switch: () => Switch
});
module.exports = __toCommonJS(Switch_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_Switch_css = require("./Switch.sty.js");
var import_VisuallyHidden = require("../../core/components/visually-hidden/VisuallyHidden.js");
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useControllableState = require("../../core/hooks/useControllableState.js");
var import_useFormControlState = require("../../core/hooks/useFormControlState.js");
var import_useFocusRing = require("../../core/styles/useFocusRing.js");
var import_is_node_environment = require("../../core/utils/is-node-environment.js");
var import_merge_props = require("../../core/utils/merge-props.js");
var import_Flex = require("../../layouts/flex/Flex.js");
var import_field_css = require("../../styles/field.sty.js");
var import_FormFieldMessageBaseItem = require("../forms-core/components/form-field-messages/FormFieldMessageBaseItem.js");
var import_useFormFieldMessagesContext = require("../forms-core/hooks/useFormFieldMessagesContext.js");
var import_useFormFieldValidation = require("../forms-core/hooks/useFormFieldValidation.js");
var import_useSyncedFormControlProps = require("../forms-core/hooks/useSyncedFormControlProps.js");
var import_formfield_messages_state_reducer = require("../forms-core/state/formfield-messages-state-reducer.js");
var import_useInputHint = require("../hooks/useInputHint.js");
const COMPONENT_NAME = "Switch";
const Switch = (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const {
      value,
      defaultValue,
      children,
      controlState,
      name,
      formValue,
      onChange,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      id: propId,
      required: propRequired,
      disabled: propDisabled,
      "aria-disabled": ariaDisabledProp,
      readOnly: propReadOnly,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
    const { state: messagesContextState, dispatch: messagesContextDispatch } = (0, import_useFormFieldMessagesContext.useFormFieldMessagesContext)();
    const { disabled, required, id, ariaDisabled, readOnly } = (0, import_useSyncedFormControlProps.useSyncedFormControlProps)({
      propDisabled,
      propRequired,
      propId,
      propAriaDisabled: ariaDisabledProp,
      propReadOnly
    });
    const disabledOrAriaDisabled = disabled || ariaDisabled || false;
    const ariaLabelingProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(props);
    const rootRef = (0, import_react.useRef)(null);
    const inputRef = (0, import_react.useRef)(null);
    (0, import_react.useImperativeHandle)(forwardedRef, () => {
      return {
        element: rootRef.current,
        inputRef: inputRef.current,
        focus() {
          inputRef.current?.focus();
        },
        validate() {
          inputRef.current?.checkValidity();
          messagesContextDispatch?.({
            type: import_formfield_messages_state_reducer.FormFieldMessagesAction.SETISTOUCHED,
            payload: { isTouched: true }
          });
        }
      };
    });
    const {
      onChange: validationOnChange,
      onBlur: validationOnBlur,
      onInvalid: validationOnInvalid,
      ariaProps
    } = (0, import_useFormFieldValidation.useFormFieldValidation)(inputRef, { required });
    const [controlStateHasError, hint] = (0, import_useFormControlState.useFormControlState)(controlState);
    const showError = controlStateHasError || messagesContextState.showError;
    const { inputProps, hintProps } = (0, import_useInputHint.useInputHint)({
      hint,
      hasError: showError
    });
    if (value !== void 0 && !onChange && (0, import_is_node_environment.isDev)()) {
      console.log(
        'Defining "value" makes the switch controlled. Set the onChange() callback and implement the toggle functionality.'
      );
    }
    const [isSelected, setIsSelected] = (0, import_useControllableState.useControllableState)({
      value,
      defaultValue
    });
    const color = isSelected ? "primary" : "neutral";
    let focusRingColor = "neutral";
    if (showError) {
      focusRingColor = "critical";
    } else if (!(readOnly || disabledOrAriaDisabled) && isSelected) {
      focusRingColor = "primary";
    }
    const { focusClassName, focusProps, isFocusVisible } = (0, import_useFocusRing.useFocusRing)({
      color: focusRingColor,
      disabled
    });
    function handleChange(e) {
      if (!disabled && !readOnly) {
        onChange?.(!isSelected, e);
        setIsSelected(!isSelected);
        !controlState && validationOnChange();
      }
    }
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      import_Flex.Flex,
      {
        ref: rootRef,
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        flexDirection: "column",
        gap: 4,
        className: consumerClassName,
        style: consumerStyle,
        ...behavioralTrackingProps,
        "data-dt-component": COMPONENT_NAME,
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
            "label",
            {
              htmlFor: id,
              className: (0, import_clsx.default)(
                (0, import_Switch_css.switchLabelCSS)({ disabled: disabledOrAriaDisabled, readOnly })
              ),
              children: [
                /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_VisuallyHidden.VisuallyHidden, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  "input",
                  {
                    id,
                    type: "checkbox",
                    role: "switch",
                    checked: isSelected,
                    "aria-disabled": ariaDisabled,
                    disabled,
                    required,
                    readOnly,
                    "aria-checked": isSelected,
                    onChange: handleChange,
                    onBlur: (event) => {
                      focusProps.onBlur(event);
                      !controlState && validationOnBlur(event);
                    },
                    onFocus: focusProps.onFocus,
                    onInvalid: (event) => {
                      !controlState && validationOnInvalid(event);
                    },
                    name,
                    value: formValue,
                    ref: inputRef,
                    ...(0, import_merge_props.mergeProps)(
                      controlState ? inputProps : ariaProps,
                      ariaLabelingProps
                    )
                  }
                ) }),
                /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  "div",
                  {
                    className: (0, import_clsx.default)(
                      focusClassName,
                      (0, import_Switch_css.switchWrapperCSS)({
                        disabled: disabledOrAriaDisabled,
                        color,
                        hasError: showError,
                        isFocusVisible,
                        readOnly
                      }),
                      (0, import_field_css.fieldTokens)({
                        disabled: disabledOrAriaDisabled,
                        color: showError ? "critical" : color,
                        variant: isSelected ? "accent" : showError ? "emphasized" : "default"
                      })
                    ),
                    children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
                      "div",
                      {
                        className: (0, import_Switch_css.switchIconCSS)({
                          selected: isSelected,
                          readOnly,
                          hasError: showError
                        }),
                        children: [
                          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.CheckmarkIcon, { size: "small" }),
                          /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: import_Switch_css.switchThumbCSS }),
                          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.XmarkIcon, { size: "small" })
                        ]
                      }
                    )
                  }
                ),
                /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  "span",
                  {
                    className: (0, import_Switch_css.switchTextCSS)({
                      disabled: disabledOrAriaDisabled,
                      required: !!children && required
                    }),
                    children
                  }
                )
              ]
            }
          ),
          hint && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_FormFieldMessageBaseItem.InternalBaseItem,
            {
              variant: showError ? "error" : "hint",
              disabled: disabledOrAriaDisabled,
              ...hintProps,
              children: hint
            }
          )
        ]
      }
    );
  }
);
Switch.displayName = COMPONENT_NAME;
