// @dynatrace/strato-components/charts/meter-bar/multi-meter-bar/utils/build-multi-meter-bar-chart-config.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var build_multi_meter_bar_chart_config_exports = {};
__export(build_multi_meter_bar_chart_config_exports, {
  buildMultiMeterBarChartConfig: () => buildMultiMeterBarChartConfig
});
module.exports = __toCommonJS(build_multi_meter_bar_chart_config_exports);
var import_lodash_es = require("lodash");
var import_constants = require("../../shared/constants.js");
var import_meter_bar_toolbar = require("../../shared/types/meter-bar-toolbar.js");
var import_get_numeric_value = require("../../shared/utils/get-numeric-value.js");
var import_default = require("../default.js");
var import_multi_meter_bar_chart = require("../types/multi-meter-bar-chart.js");
const getToolbarConfig = (toolbarSlotsConfig, toolbarConsumerConfig) => {
  const configuredToolbar = toolbarSlotsConfig ?? toolbarConsumerConfig;
  return { ...import_meter_bar_toolbar.defaultToolbarInternalConfig, ...configuredToolbar };
};
const getMaxValue = (totalValue, maxProperty, consumerConfigMax) => {
  if (typeof maxProperty === "number") {
    return maxProperty;
  }
  if (maxProperty === "auto") {
    return totalValue;
  }
  if (!maxProperty && consumerConfigMax) {
    return consumerConfigMax;
  }
  return totalValue;
};
const isValidTooltipVariant = (tooltipVariant) => import_multi_meter_bar_chart.multiMeterBarChartTooltipVariantOptions.includes(tooltipVariant);
const getTooltipConfig = (slots, consumerConfig) => {
  const consumerTooltip = consumerConfig.tooltip ? {
    ...consumerConfig.tooltip.enabled !== void 0 && {
      hidden: !consumerConfig.tooltip.enabled
    },
    variant: consumerConfig.tooltip.variant
  } : void 0;
  const merged = (0, import_lodash_es.merge)(
    {},
    import_default.defaultConfig.tooltip,
    consumerTooltip,
    slots.tooltip
  );
  const variant = isValidTooltipVariant(merged.variant) ? merged.variant : import_default.defaultConfig.tooltip.variant;
  return { ...merged, variant };
};
const buildMultiMeterBarChartConfig = (props, consumerConfig, slots) => {
  const size = props.size ?? consumerConfig.size ?? import_default.defaultConfig.size;
  const colorPalette = props.colorPalette ?? consumerConfig.colorPalette ?? import_default.defaultConfig.colorPalette;
  const min = props.min ?? consumerConfig.min ?? import_default.defaultConfig.min;
  const tooltip = getTooltipConfig(slots, consumerConfig);
  const formatter = props.formatter ?? consumerConfig.formatter ?? import_constants.DEFAULT_FORMATTER;
  const unit = props.unit ?? consumerConfig.unit;
  const legend = slots.legend ?? consumerConfig.legend ?? import_default.defaultConfig.legend;
  const totalValue = slots.segments.reduce((acc, segment) => {
    const value = (0, import_get_numeric_value.getNumericValue)(segment.value, segment.valueAccessor);
    return acc + value;
  }, 0);
  const max = getMaxValue(totalValue, props.max, consumerConfig.max);
  const colorRules = slots.colorRules ?? consumerConfig.colorRules ?? [];
  const toolbar = getToolbarConfig(slots.toolbar, consumerConfig.toolbar);
  const loading = props.loading ?? consumerConfig.loading ?? import_default.defaultConfig.loading;
  return {
    segments: slots.segments,
    colorPalette,
    size,
    min,
    max,
    loading,
    tooltip,
    labelSlot: slots.label,
    valueSlot: slots.value,
    minSlot: slots.min,
    maxSlot: slots.max,
    legend,
    formatter,
    unit,
    totalValue,
    colorRules,
    toolbar
  };
};
