// @dynatrace/strato-components/esm/charts/xy-chart/types/guards.js
import {
  DOT_VARIANT,
  LINE_VARIANT,
  AREA_VARIANT
} from "./xy-chart-internals.js";
const isNumber = (value) => {
  return typeof value === "number";
};
const isString = (value) => {
  return typeof value === "string";
};
const isDate = (value) => {
  return value instanceof Date;
};
const isFunction = (fn) => {
  return typeof fn === "function";
};
const REGEX_LINEAR = /linear/;
const REGEX_LOG = /log/;
const REGEX_TIME = /^time$/;
const REGEX_CATEGORICAL = /^categorical$/;
const REGEX_NUMERICAL = /numerical/;
const REGEX_ANY_NUMBER = /^numerical|^time$/;
const isLinear = (axisType) => REGEX_LINEAR.test(axisType);
const isLogarithmic = (axisType) => REGEX_LOG.test(axisType);
const isTime = (axisType) => REGEX_TIME.test(axisType);
const isCategorical = (axisType) => REGEX_CATEGORICAL.test(axisType);
const isNumerical = (axisType) => REGEX_NUMERICAL.test(axisType);
const isAnyNumber = (axisType) => REGEX_ANY_NUMBER.test(axisType);
function isLinearAxis(axis) {
  return isLinear(axis.type);
}
function isLogarithmicAxis(axis) {
  return isLogarithmic(axis.type);
}
function isNumericalAxis(axis) {
  return isNumerical(axis.type);
}
function isAnyNumericalAxis(axis) {
  return !isCategorical(axis.type);
}
function isTimeAxis(axis) {
  return isTime(axis.type);
}
function isCategoricalAxis(axis) {
  return isCategorical(axis.type);
}
function isCategoricalScale(args) {
  return !!args.axisType && isCategoricalAxis({ type: args.axisType });
}
function isTimeScale(args) {
  return !!args.axisType && isTime(args.axisType);
}
function isAnyNumericalScale(args) {
  return !!args.axisType && isAnyNumericalAxis({ type: args.axisType });
}
function isDotContentPoint(content) {
  return !!content && content.type === DOT_VARIANT && !!content.series && !!content.datapoint;
}
function isLineContentPoint(content) {
  return !!content && content.type === LINE_VARIANT && !!content.series && !!content.datapoint;
}
function isAreaContentPoint(content) {
  return !!content && content.type === AREA_VARIANT && !!content.series && !!content.datapoint;
}
function isNumericColoringScale(colorScale) {
  return colorScale.domain().every((value) => typeof value === "number");
}
const isValidCategoryValue = (value) => Boolean(value && typeof value === "string");
function isDotOrLineSelection(point) {
  return point.series.type === DOT_VARIANT || point.series.type === LINE_VARIANT;
}
export {
  isAnyNumber,
  isAnyNumericalAxis,
  isAnyNumericalScale,
  isAreaContentPoint,
  isCategorical,
  isCategoricalAxis,
  isCategoricalScale,
  isDate,
  isDotContentPoint,
  isDotOrLineSelection,
  isFunction,
  isLineContentPoint,
  isLinear,
  isLinearAxis,
  isLogarithmic,
  isLogarithmicAxis,
  isNumber,
  isNumericColoringScale,
  isNumerical,
  isNumericalAxis,
  isString,
  isTime,
  isTimeAxis,
  isTimeScale,
  isValidCategoryValue
};
