// @dynatrace/strato-components/esm/charts/xy-chart/types/xy-chart-internals.js
import {
  ORIENTATION_HORIZONTAL,
  ORIENTATION_VERTICAL
} from "../../core/components/canvas/types/canvas-shape.js";
const RANGE_LEGEND_TYPE = "range";
const CATEGORICAL_LEGEND_TYPE = "categorical";
const RECT_VARIANT = "rect";
const DOT_VARIANT = "dot";
const LINE_VARIANT = "line";
const BAR_VARIANT = "bar";
const AREA_VARIANT = "area";
const BAND_VARIANT = "band";
const AREA_HORIZONTAL = `${AREA_VARIANT}${ORIENTATION_HORIZONTAL}`;
const AREA_VERTICAL = `${AREA_VARIANT}${ORIENTATION_VERTICAL}`;
const BAND_HORIZONTAL = `${BAND_VARIANT}${ORIENTATION_HORIZONTAL}`;
const BAND_VERTICAL = `${BAND_VARIANT}${ORIENTATION_VERTICAL}`;
const DOT_ID = "dot_id";
const LINE_ID = "line_id";
const AREA_ID = "area_id";
const RECT_ID = "rect_id";
const BAR_ID = "bar_id";
export {
  AREA_HORIZONTAL,
  AREA_ID,
  AREA_VARIANT,
  AREA_VERTICAL,
  BAND_HORIZONTAL,
  BAND_VARIANT,
  BAND_VERTICAL,
  BAR_ID,
  BAR_VARIANT,
  CATEGORICAL_LEGEND_TYPE,
  DOT_ID,
  DOT_VARIANT,
  LINE_ID,
  LINE_VARIANT,
  RANGE_LEGEND_TYPE,
  RECT_ID,
  RECT_VARIANT
};
