// @dynatrace-sdk/navigation/testing/cjs/index.js
/**
 * @license
 * Copyright 2023-2025 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/platform/navigation/testing/index.ts
var testing_exports = {};
__export(testing_exports, {
  clearNavigationMocks: () => clearNavigationMocks,
  mockNavigation: () => mockNavigation,
  setupTestingEnv: () => setupTestingEnv
});
module.exports = __toCommonJS(testing_exports);

// packages/platform/shared/utils/types-guards.ts
function isString(val) {
  return typeof val === "string";
}

// packages/platform/navigation/testing/test-utils.ts
var currentSetup;
var isInitialised = false;
var wrapper = {
  getIntent: () => currentSetup.getIntent(),
  getIntentLink: (intentPayload, appId, intentId) => {
    if (appId !== void 0 && intentId !== void 0) {
      return currentSetup.getIntentLink(intentPayload, appId, intentId);
    }
    return currentSetup.getIntentLink(intentPayload);
  },
  sendIntent: (intentPayload, legacyRecommendedAppIdOrSendIntentOptions, recommendedIntentId) => {
    if (isString(legacyRecommendedAppIdOrSendIntentOptions) && isString(recommendedIntentId)) {
      return currentSetup.sendIntent(
        intentPayload,
        legacyRecommendedAppIdOrSendIntentOptions,
        recommendedIntentId
      );
    } else if (legacyRecommendedAppIdOrSendIntentOptions !== void 0) {
      const sendIntentOptions = legacyRecommendedAppIdOrSendIntentOptions;
      return currentSetup.sendIntent(intentPayload, sendIntentOptions);
    }
    return currentSetup.sendIntent(intentPayload);
  },
  sendIntentWithResponse: (intentPayload, sendIntentWithResponseOptions) => currentSetup.sendIntentWithResponse(intentPayload, sendIntentWithResponseOptions),
  sendIntentWithResponseV2: (intentPayload, sendIntentWithResponseOptions) => currentSetup.sendIntentWithResponseV2(intentPayload, sendIntentWithResponseOptions),
  getAppLink: (appId, pageToken) => currentSetup.getAppLink(appId, pageToken),
  openApp: (appId, pageToken) => currentSetup.openApp(appId, pageToken),
  setPathChangeHandler: (handler) => {
    return currentSetup.setPathChangeHandler(handler);
  },
  getDocumentLink: (documentId) => currentSetup.getDocumentLink(documentId),
  openDocument: (documentId) => currentSetup.openDocument(documentId)
};
var setupTestingEnv = () => {
  if (typeof expect === "undefined") {
    isInitialised = true;
  }
  mockNavigation();
};
var mockNavigation = (setup = {}) => {
  if (!isInitialised) {
    throw new Error(`\`@dynatrace-sdk/navigation/testing\` has not been initialised!
Add in your \`jest.config.js\` file entry with \`@dynatrace-sdk/navigation/testing\` setup:

\`setupFiles: ['@dynatrace-sdk/navigation/testing']\``);
  }
  const globalWithDtRuntimeNavigation = globalThis;
  if (globalWithDtRuntimeNavigation.dtRuntime === void 0) {
    globalWithDtRuntimeNavigation.dtRuntime = {};
  }
  if (globalWithDtRuntimeNavigation.dtRuntime.navigation === void 0) {
    globalWithDtRuntimeNavigation.dtRuntime.navigation = wrapper;
  }
  const legacyDtSdk = globalThis;
  legacyDtSdk.dtSdk = legacyDtSdk.dtSdk || {};
  legacyDtSdk.dtSdk.getIntent = wrapper.getIntent;
  legacyDtSdk.dtSdk.getIntentLink = wrapper.getIntentLink;
  legacyDtSdk.dtSdk.sendIntent = wrapper.sendIntent;
  legacyDtSdk.dtSdk.openApp = wrapper.openApp;
  legacyDtSdk.dtSdk.getAppLink = wrapper.getAppLink;
  legacyDtSdk.dtSdk.setPathChangeHandler = wrapper.setPathChangeHandler;
  currentSetup = getSetupWithDefaults(setup);
};
var clearNavigationMocks = () => {
  mockNavigation();
};
var getSetupWithDefaults = (setup) => {
  return {
    getDocumentLink: setup.getDocumentLink || (() => "mocked-document-link"),
    openDocument: setup.openDocument || (() => null),
    sendIntent: setup.sendIntent || (() => null),
    sendIntentWithResponse: async () => ({
      response: { "mocked.property": "mocked-value" }
    }),
    sendIntentWithResponseV2: setup.sendIntentWithResponseV2 || (() => Promise.resolve({
      ok: true,
      data: { "mocked.property": "mocked-value" }
    })),
    getIntent: setup.getIntent || (() => null),
    getIntentLink: setup.getIntentLink || (() => "mocked-intent-link"),
    getAppLink: setup.getAppLink || (() => "mocked-app-link"),
    openApp: setup.openApp || (() => null),
    setPathChangeHandler: setup.setPathChangeHandler || (() => null)
  };
};

// packages/platform/navigation/testing/index.ts
setupTestingEnv();
