// @dynatrace/strato-components/esm/charts/xy-chart/hooks/useXYChartDataProvider.js
import { useIntl } from "react-intl";
import { useAxesConfig } from "./useAxesConfig.js";
import { useColorPalette } from "./useColorPalette.js";
import { useSortLinearDatapoints } from "./useSortLinearDatapoints.js";
import { useSortRectangleDatapoints } from "./useSortRectangleDatapoints.js";
import {
  getStrategyFactory
} from "../processing-strategies/factory/factory.js";
import { AreaSeriesStrategy } from "../processing-strategies/strategies/area-series.js";
import { BarSeriesStrategy } from "../processing-strategies/strategies/bar-series.js";
import { DotSeriesStrategy } from "../processing-strategies/strategies/dot-series.js";
import { LineSeriesStrategy } from "../processing-strategies/strategies/line-series.js";
import { RectSeriesStrategy } from "../processing-strategies/strategies/rect-series.js";
import { getDatumValue } from "../processing-strategies/utils/access-datapoint/get-datapoints.js";
import { useFilteredSeriesIdsState } from "../store/selectors.js";
import { isCategorical } from "../types/guards.js";
import {
  isInfiniteZoomEnabledFor,
  isPanEnabledFor,
  isZoomEnabledFor
} from "../utils/axis-interactions.js";
import { combineDataWithCorrespondingSeries } from "../utils/combine-data-with-corresponding-series.js";
import { combineSeriesWithAxes } from "../utils/combine-series-with-axes.js";
function useXYChartDataProvider(seriesVariantsProps, data) {
  const axes = useAxesConfig();
  const colorPalette = useColorPalette();
  const intl = useIntl();
  const axisIdToScaleDomain = /* @__PURE__ */ new Map();
  const axisWithSeries = /* @__PURE__ */ new Set();
  const datapointMetadataMap = /* @__PURE__ */ new Map();
  const axisIdToCategoryDatum = /* @__PURE__ */ new Map();
  const {
    addValue,
    getCurrentSelectConfig: getCurrentShapeSelectConfig,
    searchAllAxisDatapoints: searchAllAxisShapeDatapoints
  } = useSortLinearDatapoints(seriesVariantsProps, datapointMetadataMap);
  const {
    addValue: addRectangleValue,
    searchAllAxisDatapoints: searchAllAxisRectangleDatapoints,
    getCurrentSelectConfig: getCurrentRectangleSelectConfig
  } = useSortRectangleDatapoints(seriesVariantsProps);
  const seriesWithAxes = combineSeriesWithAxes(seriesVariantsProps, axes);
  const dataWithSeries = combineDataWithCorrespondingSeries(
    data,
    seriesWithAxes
  );
  const seriesWithDatapoints = [];
  const seriesToMetadata = /* @__PURE__ */ new Map();
  const seriesMap = /* @__PURE__ */ new Map();
  const seriesIndexes = [];
  const rectSeriesStrategy = new RectSeriesStrategy({
    addValue: addRectangleValue
  });
  const dotSeriesStrategy = new DotSeriesStrategy({ addValue });
  const lineSeriesStrategy = new LineSeriesStrategy({ addValue });
  const barSeriesStrategy = new BarSeriesStrategy({
    addValue: addRectangleValue
  });
  const areaSeriesStrategy = new AreaSeriesStrategy({ addValue });
  const strategies = [
    rectSeriesStrategy,
    dotSeriesStrategy,
    lineSeriesStrategy,
    barSeriesStrategy,
    areaSeriesStrategy
  ];
  const filteredSeriesIds = useFilteredSeriesIdsState();
  const mutableContext = {
    seriesMap,
    seriesWithDatapoints,
    seriesToMetadata,
    axisIdToScaleDomain,
    seriesIndexes,
    intl,
    colorPalette,
    datapointMetadataMap,
    filteredSeriesIds,
    axisWithSeries
  };
  const captureCategoryDatum = (axis, accessor, datum) => {
    if (!isCategorical(axis.type) || !accessor) {
      return;
    }
    const rawValue = getDatumValue(datum, accessor);
    if (typeof rawValue !== "string" && typeof rawValue !== "number") {
      return;
    }
    const raw = String(rawValue);
    let perAxis = axisIdToCategoryDatum.get(axis.id);
    if (!perAxis) {
      perAxis = /* @__PURE__ */ new Map();
      axisIdToCategoryDatum.set(axis.id, perAxis);
    }
    if (!perAxis.has(raw)) {
      perAxis.set(raw, datum);
    }
  };
  for (const { data: data2, series } of dataWithSeries.filter(
    ({ series: series2 }) => series2.length
  )) {
    for (const { seriesProps, xAxis, yAxis } of series) {
      const strategy = getStrategyFactory(strategies, seriesProps);
      if (strategy?.preprocess) {
        strategy.preprocess(data2, seriesProps, xAxis, yAxis, mutableContext);
      }
    }
    for (const datum of data2) {
      for (const { seriesProps, xAxis, yAxis } of series) {
        const strategy = getStrategyFactory(strategies, seriesProps);
        if (strategy) {
          const processResult = strategy.process(
            datum,
            seriesProps,
            xAxis,
            yAxis,
            mutableContext
          );
          if (processResult === "continue") {
            continue;
          }
        } else {
          throw new Error(
            `No strategy found to process series with id: ${seriesProps.id}`
          );
        }
        captureCategoryDatum(xAxis, seriesProps.x0Accessor, datum);
        captureCategoryDatum(yAxis, seriesProps.y0Accessor, datum);
      }
    }
  }
  const validAxes = axes.filter(
    (axis) => axisIdToScaleDomain.has(axis.id) && axisWithSeries.has(axis.id)
  );
  const linearSelectionContextValue = {
    selectionConfigByAxisId: getCurrentShapeSelectConfig(),
    searchAllAxisDatapoints: searchAllAxisShapeDatapoints
  };
  const rectSelectionContextValue = {
    selectionConfigByAxisId: getCurrentRectangleSelectConfig(),
    searchAllAxisDatapoints: searchAllAxisRectangleDatapoints
  };
  return {
    axesMetadata: {
      isZoomEnabled: !!validAxes.find(isZoomEnabledFor),
      isPanEnabled: !!validAxes.find(isPanEnabledFor),
      isInfinite: !!validAxes.find(isInfiniteZoomEnabledFor)
    },
    axes: validAxes,
    axisIdToScaleDomain,
    axisIdToCategoryDatum,
    seriesIdToValueScaleDomain: rectSeriesStrategy.seriesIdToValueScaleDomain,
    datapointMetadataMap,
    linearSelectionContextValue,
    rectSelectionContextValue,
    seriesWithDatapoints,
    filteredSeriesWithDatapoints: seriesWithDatapoints.filter((series) => {
      if (series.type === "rect") {
        series.filteredData = series.data.filter(
          (datapoint) => {
            return datapointMetadataMap.get(datapoint)?.isVisible !== false;
          }
        );
      }
      return seriesToMetadata.get(series)?.isVisible !== false;
    }),
    seriesToMetadata,
    axisIdToDimension: barSeriesStrategy.axisIdToDimension
  };
}
export {
  useXYChartDataProvider
};
