// @dynatrace/strato-components/esm/charts/timeseries/components/overlays/TimeseriesExplore/TimeseriesChartTooltip.js
import { jsx, jsxs } from "react/jsx-runtime";
import { ChartActionsMenu } from "../../../../core/components/series-action/ChartActionsMenu.js";
import { useChartActionsMenu } from "../../../../core/components/series-action/hooks/useChartActionsMenu.js";
import { getDefaultAction } from "../../../../core/components/series-action/hooks/useDefaultAction.js";
import { InteractiveChartTooltip } from "../../../../core/components/tooltip/InteractiveChartTooltip/components/InteractiveChartTooltip.js";
import { SeriesBasedContent } from "../../../../core/components/tooltip/InteractiveChartTooltip/contents/SeriesBasedContent.js";
import { TimebasedTooltipHeader } from "../../../../core/components/tooltip/InteractiveChartTooltip/headers/TimebasedTooltipHeader.js";
import { useSeriesBasedItemEstimatedSize } from "../../../../core/components/tooltip/InteractiveChartTooltip/hooks/useSeriesBasedItemEstimatedSize.js";
import { useTooltipSeriesActions } from "../../../../core/components/tooltip/shared/hooks/useTooltipSeriesActions.js";
import { useScrollResponsiveChartBounds } from "../../../../core/hooks/useScrollResponsiveChartBounds.js";
import { calculateAdjustedTooltipPosition } from "../../../../core/utils/calculate-adjusted-tooltip-position.js";
import { safeGet } from "../../../../core/utils/safe-get.js";
import { useExploreState } from "../../../hooks/store/selectors.js";
import { useTimeseriesFilteredData } from "../../../hooks/useTimeseriesFilteredData.js";
import { useTimeseriesOverlayRef } from "../../../hooks/useTimeseriesOverlayRef.js";
import { useTimeseriesUnits } from "../../../hooks/useTimeseriesUnits.js";
import { useTimeseriesYFormatter } from "../../../hooks/useTimeseriesYFormatter.js";
import { useTooltipConfig } from "../../../hooks/useTooltipConfig.js";
import { useTruncationMode } from "../../../hooks/useTruncationMode.js";
import { ADJUSTED_X_AXIS, SHARED_TOOLTIP_VARIANT } from "../constants.js";
import { useSeriesActionClickHandler } from "../hooks/useSeriesActionClickHandler.js";
import { sortTooltipItems } from "../utils/tooltip/sort-tooltip-items.js";
import {
  filterGroupedPoints,
  toTooltipItems
} from "../utils/tooltip/timeseries-tooltip.js";
const TooltipPositionWrapper = ({
  tooltipItems,
  onMouseLeave,
  onScrollContent,
  children
}) => {
  const { pinned, visible, position } = useExploreState();
  const chartRef = useTimeseriesOverlayRef();
  const bounds = useScrollResponsiveChartBounds(chartRef, pinned)();
  const { seriesDisplayMode } = useTooltipConfig();
  const estimatedItemSize = useSeriesBasedItemEstimatedSize(
    tooltipItems,
    seriesDisplayMode
  );
  const adjustedTooltipPosition = calculateAdjustedTooltipPosition(
    bounds,
    position,
    { x: ADJUSTED_X_AXIS, y: 0 }
  );
  const showTooltip = tooltipItems.length && visible && adjustedTooltipPosition;
  return showTooltip ? /* @__PURE__ */ jsx(
    InteractiveChartTooltip,
    {
      position: adjustedTooltipPosition,
      pinned,
      itemSize: estimatedItemSize,
      updatePositionWhenPinned: true,
      onMouseLeave,
      onScrollContent,
      children
    }
  ) : null;
};
const TimeseriesChartTooltip = ({
  series,
  closestPoint,
  xPoints,
  groupedPoints
}) => {
  const { seriesDisplayMode, variant } = useTooltipConfig();
  const truncationMode = useTruncationMode();
  const { seriesMap, datapointsMap } = useTimeseriesFilteredData();
  const {
    onLeave: handleMenuLeave,
    onScroll: handleMenuScroll,
    selectedItem,
    isScrolling,
    updateSelectedItem,
    actionsMenuRef
  } = useChartActionsMenu();
  const formatters = useTimeseriesYFormatter();
  const units = useTimeseriesUnits();
  const getTooltipTimeframe = () => [
    closestPoint.datapoint.start,
    closestPoint.datapoint.end
  ];
  const pointsToDisplay = variant === SHARED_TOOLTIP_VARIANT ? sortTooltipItems(xPoints) : sortTooltipItems(
    filterGroupedPoints(closestPoint, xPoints, groupedPoints)
  );
  const tooltipItems = toTooltipItems(
    pointsToDisplay,
    formatters,
    datapointsMap,
    units
  );
  const closestItem = tooltipItems.find((item) => item.id === closestPoint.id);
  const actionClickHandler = useSeriesActionClickHandler(
    tooltipItems.length > 1
  );
  const actions = useTooltipSeriesActions(
    series,
    (series2) => safeGet(seriesMap, series2).id,
    tooltipItems.find((t) => t.id === selectedItem)?.datapoint
  );
  const isShared = variant === SHARED_TOOLTIP_VARIANT;
  return /* @__PURE__ */ jsxs(
    TooltipPositionWrapper,
    {
      tooltipItems,
      onMouseLeave: handleMenuLeave,
      onScrollContent: handleMenuScroll,
      children: [
        /* @__PURE__ */ jsx(InteractiveChartTooltip.Header, { children: /* @__PURE__ */ jsx(
          TimebasedTooltipHeader,
          {
            timeframe: getTooltipTimeframe(),
            item: isShared && xPoints.length > 1 ? closestItem : void 0,
            seriesDisplayMode
          }
        ) }),
        /* @__PURE__ */ jsx(InteractiveChartTooltip.Body, { children: tooltipItems.map((item, index) => () => {
          const isOpen = selectedItem === item.id;
          return /* @__PURE__ */ jsx(
            ChartActionsMenu,
            {
              isOpen,
              actions: isOpen ? actions(item) : void 0,
              defaultActions: getDefaultAction(item),
              ref: actionsMenuRef,
              onActionClick: actionClickHandler,
              children: /* @__PURE__ */ jsx(
                SeriesBasedContent,
                {
                  item: { ...item, values: item.asyncValues() },
                  onMouseHover: () => updateSelectedItem(item.id),
                  seriesDisplayMode,
                  showDivider: index !== tooltipItems.length - 1,
                  truncationMode,
                  isScrolling
                }
              )
            },
            `${item.id}`
          );
        }) })
      ]
    }
  );
};
export {
  TimeseriesChartTooltip
};
