// @dynatrace/strato-components/esm/charts/timeseries/components/overlays/TimeseriesExplore/TimeseriesCrosshair.js
import { jsx, jsxs } from "react/jsx-runtime";
import { useGridPlotSize } from "../../../../core/components/chart-grid/hooks/useGridPlotSize.js";
import { HorizontalCrosshair } from "../../../../core/components/crosshair/HorizontalCrosshair.js";
import { VerticalCrosshair } from "../../../../core/components/crosshair/VerticalCrosshair.js";
import { OVERLAY_PADDING } from "../../../../core/components/data-point/constants.js";
import { reduceResetExploreVisible } from "../../../../core/components/interactions-state-helpers/reducers/reset-reducers.js";
import { THRESHOLD_BAR_THICKNESS } from "../../../../core/components/threshold-bar/types/constants.js";
import { useApplyCallbackOnScroll } from "../../../../core/hooks/useApplyCallbackOnScroll.js";
import {
  useExploreState,
  useSetState
} from "../../../hooks/store/selectors.js";
import { useThresholdConfig } from "../../../hooks/useThresholdConfig.js";
import { useTimeseriesFilteredData } from "../../../hooks/useTimeseriesFilteredData.js";
import { useTimeseriesOverlayRef } from "../../../hooks/useTimeseriesOverlayRef.js";
import { useTimeseriesXFormatter } from "../../../hooks/useTimeseriesXFormatter.js";
import { useTimeseriesYFormatter } from "../../../hooks/useTimeseriesYFormatter.js";
import { useXScales } from "../../../hooks/useXScale.js";
import { useYScales } from "../../../hooks/useYScales.js";
import { crosshairLayout } from "../TimeseriesOverlays.sty.js";
import {
  generateXMagnifierLabels,
  generateYMagnifierLabels
} from "../utils/magnifiers-utils.js";
function TimeseriesCrosshair({
  closestPoint
}) {
  const { units } = useTimeseriesFilteredData();
  const yFormatters = useTimeseriesYFormatter();
  const xFormatter = useTimeseriesXFormatter();
  const yScale = useYScales();
  const { currentScale: xScale } = useXScales();
  const { width, height } = useGridPlotSize();
  const overlaysRef = useTimeseriesOverlayRef();
  const { pinned, visible, position } = useExploreState();
  const { left: leftThreshold, right: rightThreshold } = useThresholdConfig();
  const hasLeftThresholds = leftThreshold && leftThreshold.length > 0;
  const hasRightThresholds = rightThreshold && rightThreshold.length > 0;
  const relativeMousePosition = position?.relative;
  const chartBounds = overlaysRef?.current?.getBoundingClientRect();
  const isMouseHovering = relativeMousePosition && visible;
  const showVerticalCrosshair = isMouseHovering || pinned;
  const showHorizontalCrosshair = isMouseHovering && !pinned;
  const thresholdLeftSize = hasLeftThresholds ? THRESHOLD_BAR_THICKNESS : 0;
  const thresholdRightSize = hasRightThresholds ? THRESHOLD_BAR_THICKNESS : 0;
  const setState = useSetState();
  const isCallbackOnScrollDisabled = pinned || !isMouseHovering;
  const onScrollCallback = () => setState(reduceResetExploreVisible);
  useApplyCallbackOnScroll(isCallbackOnScrollDisabled, onScrollCallback);
  const horizontalCrosshairWidth = width + thresholdLeftSize + thresholdRightSize;
  const verticalXCrosshair = pinned && closestPoint ? closestPoint.x - OVERLAY_PADDING : relativeMousePosition?.x ?? 0;
  const yMagnifierLabels = relativeMousePosition ? generateYMagnifierLabels(
    yScale,
    relativeMousePosition,
    units,
    yFormatters
  ) : void 0;
  const hoveredDate = xScale.invert(relativeMousePosition?.x ?? 0);
  const xMagnifierLabel = pinned ? "" : generateXMagnifierLabels(xFormatter, hoveredDate);
  return /* @__PURE__ */ jsxs("svg", { className: crosshairLayout, children: [
    showHorizontalCrosshair && /* @__PURE__ */ jsx(
      HorizontalCrosshair,
      {
        leftLabel: yMagnifierLabels?.left,
        rightLabel: yMagnifierLabels?.right,
        y: relativeMousePosition?.y,
        width: horizontalCrosshairWidth,
        height,
        leftOffset: thresholdLeftSize,
        rightOffset: thresholdRightSize,
        chartBounds
      }
    ),
    showVerticalCrosshair && /* @__PURE__ */ jsx(
      VerticalCrosshair,
      {
        x: verticalXCrosshair,
        bottomLabel: xMagnifierLabel,
        height,
        width: horizontalCrosshairWidth,
        dashed: !pinned,
        chartBounds
      }
    )
  ] });
}
export {
  TimeseriesCrosshair
};
