// @dynatrace/strato-components/typography/text-ellipsis/TextEllipsis.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TextEllipsis_exports = {};
__export(TextEllipsis_exports, {
  TextEllipsis: () => TextEllipsis
});
module.exports = __toCommonJS(TextEllipsis_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_use_resize_observer = __toESM(require("use-resize-observer"));
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useFontsUpdated = require("../../core/hooks/useFontsUpdated.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_useSafeForwardProps = require("../../core/hooks/useSafeForwardProps.js");
var import_utils = require("../utils.js");
var styles = __toESM(require("./TextEllipsis.sty.js"));
const COMPONENT_NAME = "TextEllipsis";
function getFontProps(innerSpanElement) {
  const computedStyle = window.getComputedStyle(innerSpanElement, null);
  return {
    fontStyle: computedStyle.getPropertyValue("font-style"),
    fontWeight: computedStyle.getPropertyValue("font-weight"),
    fontSize: computedStyle.getPropertyValue("font-size"),
    fontFamily: computedStyle.getPropertyValue("font-family")
  };
}
function isNativeEllipsisActive(innerSpanElement, containerWidth) {
  const originalTextWidth = Math.max(
    innerSpanElement.getBoundingClientRect().width,
    innerSpanElement.scrollWidth
  );
  const textWidth = Math.floor(originalTextWidth * 100) / 100;
  return textWidth > containerWidth;
}
function resetFullText(innerSpanElement, originalText) {
  if (innerSpanElement.textContent !== originalText) {
    innerSpanElement.textContent = originalText;
  }
}
function getWidth(element) {
  return Math.max(element.getBoundingClientRect().width, element.clientWidth);
}
function getContainerWidth(containerElement) {
  return Math.ceil(getWidth(containerElement) * 100) / 100;
}
function onTextOverflowCallback(current, wasEllipsized, onTextOverflow) {
  if (onTextOverflow && current !== wasEllipsized) {
    onTextOverflow(current);
  }
}
function renderEllipsizedText(originalText, truncationMode, containerRef, wasEllipsized, onTextOverflow) {
  const containerElement = containerRef.current;
  const innerSpanElement = containerElement?.firstElementChild || null;
  if (containerElement === null || innerSpanElement === null) {
    return false;
  }
  resetFullText(innerSpanElement, originalText);
  const containerWidth = getContainerWidth(containerElement);
  if (!isNativeEllipsisActive(innerSpanElement, containerWidth)) {
    onTextOverflowCallback(false, wasEllipsized, onTextOverflow);
    return false;
  }
  if (truncationMode === "middle") {
    innerSpanElement.textContent = (0, import_utils.centerEllipsizeText)(
      originalText,
      getFontProps(innerSpanElement),
      containerWidth
    );
  }
  onTextOverflowCallback(true, wasEllipsized, onTextOverflow);
  return true;
}
const ComputedTextEllipsis = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    children,
    truncationMode = "end",
    onTextOverflow,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const ariaLabelingProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(restProps);
  const forwardProps = (0, import_useSafeForwardProps.useSafeForwardProps)(restProps, ariaLabelingProps);
  const { ref: observerContainerRef, width } = (0, import_use_resize_observer.default)();
  const containerRef = (0, import_react.useRef)(null);
  const ellipsisStateRef = (0, import_react.useRef)(false);
  const mergedRef = (0, import_useMergeRefs.useMergeRefs)([
    observerContainerRef,
    (instance) => {
      containerRef.current = instance;
    },
    forwardedRef
  ]);
  const fontsUpdated = (0, import_useFontsUpdated.useFontsUpdated)();
  (0, import_react.useLayoutEffect)(() => {
    ellipsisStateRef.current = renderEllipsizedText(
      children,
      truncationMode,
      containerRef,
      ellipsisStateRef.current,
      onTextOverflow
    );
  }, [width, children, truncationMode, fontsUpdated, onTextOverflow]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "span",
    {
      ref: mergedRef,
      className: (0, import_clsx.default)(
        styles.ellipsis({ truncationMode }),
        {
          [styles.active]: ellipsisStateRef.current
        },
        consumerClassName
      ),
      style: consumerStyle,
      "data-sourcetext": children,
      "data-ellipsis": true,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...ariaLabelingProps,
      ...behavioralTrackingProps,
      ...forwardProps,
      "data-dt-component": COMPONENT_NAME,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { children })
    }
  );
});
const CssTextEllipsis = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    children,
    truncationMode = "end",
    onTextOverflow,
    className: consumerClassName,
    style: consumerStyle,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "data-testid": dataTestId,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const ariaLabelingProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(restProps);
  const forwardProps = (0, import_useSafeForwardProps.useSafeForwardProps)(restProps, ariaLabelingProps);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "span",
    {
      ref: forwardedRef,
      "data-ellipsis": true,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      className: (0, import_clsx.default)(styles.ellipsis({ truncationMode }), consumerClassName),
      style: consumerStyle,
      ...ariaLabelingProps,
      ...behavioralTrackingProps,
      ...forwardProps,
      "data-dt-component": COMPONENT_NAME,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { children })
    }
  );
});
const TextEllipsis = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const cssOnlyEllipsis = !props.onTextOverflow && props.truncationMode !== "middle";
  return cssOnlyEllipsis ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(CssTextEllipsis, { ref: forwardedRef, ...props }) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(ComputedTextEllipsis, { ref: forwardedRef, ...props });
});
TextEllipsis.displayName = COMPONENT_NAME;
