// @dynatrace/strato-geo/esm/map/components/DefaultTooltip/DefaultTooltip.js
import { jsx, jsxs } from "react/jsx-runtime";
import { isNil } from "lodash-es";
import {
  _OverlayTooltip as OverlayTooltip,
  _useDefaultConfig as useDefaultConfig,
  _useDefaultLabels as useDefaultLabels
} from "@dynatrace/strato-components/charts";
import Colors from "@dynatrace/strato-design-tokens/colors";
import { LocationMarkerFilledIcon } from "@dynatrace/strato-icons";
import { useTruncationMode } from "../../hooks/use-truncation-mode.js";
import { isConnection, isRegion } from "../../utils/tooltip-type-guards.js";
const isChoroplethTooltip = (props) => isRegion(props.closestPoint);
const isConnectionTooltip = (props) => isConnection(props.closestPoint);
const isBubbleTooltip = (props) => !isNil(props.layerData);
const isDotTooltip = (props) => true;
const DefaultTooltip = (props) => {
  const defaultConfig = useDefaultConfig();
  const getLabel = useDefaultLabels();
  const truncationMode = useTruncationMode();
  let locations = [];
  if (isChoroplethTooltip(props)) {
    const { closestPoint, seriesActions } = props;
    return /* @__PURE__ */ jsxs(OverlayTooltip, { truncation: truncationMode, children: [
      /* @__PURE__ */ jsx(OverlayTooltip.SeriesHeader, { valueLabel: "" }),
      /* @__PURE__ */ jsx(
        OverlayTooltip.SeriesItem,
        {
          color: closestPoint.color,
          name: closestPoint.name,
          value: closestPoint.data.value?.toString() ?? "",
          actions: seriesActions ? () => seriesActions(closestPoint) : void 0
        }
      )
    ] });
  } else if (isConnectionTooltip(props)) {
    locations = props.closestPoint.data.path;
  } else if (isBubbleTooltip(props)) {
    const { closestPoint, layerData } = props;
    locations = layerData === void 0 || layerData === null || layerData.length === 0 ? [closestPoint.data] : layerData.map((point) => point.data);
  } else if (isDotTooltip(props)) {
    locations = [props.closestPoint.data];
  }
  const seriesLabel = getLabel?.("seriesLabel") ?? "";
  const variant = defaultConfig?.sharedVariant ? "shared" : "single";
  const seriesData = locations.map((location) => ({
    name: `${location.latitude}, ${location.longitude}`,
    value: ""
  }));
  const getActionsForItem = (current) => {
    if (props.actionsForItem) {
      return props.actionsForItem(current);
    }
    const { seriesActions: actions } = props;
    if (!actions || !locations.length) {
      return void 0;
    }
    const location = locations.find(
      (l) => `${l.latitude}, ${l.longitude}` === current.name
    );
    return location ? () => actions(location) : void 0;
  };
  return /* @__PURE__ */ jsxs(OverlayTooltip, { truncation: truncationMode, children: [
    /* @__PURE__ */ jsx(OverlayTooltip.Header, { header1: seriesLabel }),
    /* @__PURE__ */ jsx(OverlayTooltip.SeriesHeader, { disabled: true }),
    /* @__PURE__ */ jsx(
      OverlayTooltip.Series,
      {
        activeItem: 0,
        data: seriesData,
        variant,
        seriesItemTemplate: (current) => /* @__PURE__ */ jsx(
          OverlayTooltip.SeriesItem,
          {
            ...current,
            iconStyle: "large",
            actions: getActionsForItem(current),
            icon: /* @__PURE__ */ jsx(
              LocationMarkerFilledIcon,
              {
                style: {
                  fill: Colors.Text.Neutral.Default
                }
              }
            )
          }
        )
      }
    )
  ] });
};
DefaultTooltip["displayName"] = "DefaultTooltip";
export {
  DefaultTooltip
};
