// @dynatrace/strato-components/tables/DataTable/contexts/DataTableContext.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DataTableContext_exports = {};
__export(DataTableContext_exports, {
  DataTableCellContextProvider: () => DataTableCellContextProvider,
  DataTableColumnContextProvider: () => DataTableColumnContextProvider,
  DataTableColumnSettingsContextProvider: () => DataTableColumnSettingsContextProvider,
  DataTableContentContext: () => DataTableContentContext,
  DataTableContentContextProvider: () => DataTableContentContextProvider,
  DataTableContextProvider: () => DataTableContextProvider,
  DataTableHeaderContext: () => DataTableHeaderContext,
  DataTableHeaderContextProvider: () => DataTableHeaderContextProvider,
  DataTableIsHeaderContextProvider: () => DataTableIsHeaderContextProvider,
  DataTableRowContextProvider: () => DataTableRowContextProvider,
  useDataTable: () => useDataTable,
  useDataTableCell: () => useDataTableCell,
  useDataTableColumn: () => useDataTableColumn,
  useDataTableColumnSettings: () => useDataTableColumnSettings,
  useDataTableContent: () => useDataTableContent,
  useDataTableHeader: () => useDataTableHeader,
  useDataTableIsHeaderContextProvider: () => useDataTableIsHeaderContextProvider,
  useDataTableRow: () => useDataTableRow
});
module.exports = __toCommonJS(DataTableContext_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
const DataTableContext = (0, import_react.createContext)(null);
function DataTableContextProvider(props) {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(DataTableContext.Provider, { value: props.value, children: props.children });
}
function useDataTable() {
  const table = (0, import_react.useContext)(DataTableContext);
  if (table === null) {
    throw new Error("Table instance is not defined in context.");
  }
  return table;
}
const DEFAULT_CONTENT_CONTEXT = {
  hasOverflowLeft: false,
  hasOverflowRight: false,
  cellActionAriaLabel: ""
};
const DataTableContentContext = (0, import_react.createContext)(DEFAULT_CONTENT_CONTEXT);
function DataTableContentContextProvider(props) {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    DataTableContentContext.Provider,
    {
      value: props.value ?? DEFAULT_CONTENT_CONTEXT,
      children: props.children
    }
  );
}
function useDataTableContent() {
  const content = (0, import_react.useContext)(
    DataTableContentContext
  );
  if (content === null) {
    throw new Error("Content instance is not defined in context.");
  }
  return content;
}
const DataTableHeaderContext = (0, import_react.createContext)(null);
function DataTableHeaderContextProvider(props) {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(DataTableHeaderContext.Provider, { value: props.value ?? null, children: props.children });
}
function useDataTableHeader() {
  const header = (0, import_react.useContext)(
    DataTableHeaderContext
  );
  if (header === null) {
    throw new Error("Header instance is not defined in context.");
  }
  return header;
}
const DataTableColumnContext = (0, import_react.createContext)(
  null
);
function DataTableColumnContextProvider(props) {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(DataTableColumnContext.Provider, { value: props.value ?? null, children: props.children });
}
function useDataTableColumn() {
  const column = (0, import_react.useContext)(
    DataTableColumnContext
  );
  if (column === null) {
    throw new Error("Column object is not defined in context.");
  }
  return column;
}
const DataTableRowContext = (0, import_react.createContext)(null);
function DataTableRowContextProvider(props) {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(DataTableRowContext.Provider, { value: props.value ?? null, children: props.children });
}
function useDataTableRow() {
  const row = (0, import_react.useContext)(DataTableRowContext);
  return row;
}
const DataTableCellContext = (0, import_react.createContext)(null);
function DataTableCellContextProvider(props) {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(DataTableCellContext.Provider, { value: props.value ?? null, children: props.children });
}
function useDataTableCell() {
  const cell = (0, import_react.useContext)(DataTableCellContext);
  return cell;
}
const DataTableColumnSettingsContext = (0, import_react.createContext)(null);
function DataTableColumnSettingsContextProvider(props) {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(DataTableColumnSettingsContext.Provider, { value: props.value ?? null, children: props.children });
}
function useDataTableColumnSettings() {
  return (0, import_react.useContext)(
    DataTableColumnSettingsContext
  );
}
const DataTableIsHeaderContext = (0, import_react.createContext)(false);
function DataTableIsHeaderContextProvider(props) {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(DataTableIsHeaderContext.Provider, { value: true, children: props.children });
}
function useDataTableIsHeaderContextProvider() {
  return (0, import_react.useContext)(DataTableIsHeaderContext);
}
