// @dynatrace/strato-components/charts/pie/types/default.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var default_exports = {};
__export(default_exports, {
  DEFAULT_CONTAINER_PADDING: () => DEFAULT_CONTAINER_PADDING,
  DEFAULT_LEGEND: () => DEFAULT_LEGEND,
  DEFAULT_SIZE: () => DEFAULT_SIZE,
  DEFAULT_TOOLTIP_MARGIN: () => DEFAULT_TOOLTIP_MARGIN,
  DONUT_INTERNAL_RADIUS: () => DONUT_INTERNAL_RADIUS,
  LABEL_FONT_SIZE: () => LABEL_FONT_SIZE,
  LABEL_FONT_SIZE_MULTIPLIER: () => LABEL_FONT_SIZE_MULTIPLIER,
  LABEL_TEXT_POSITION: () => LABEL_TEXT_POSITION,
  PIE_CATEGORICAL_LEGEND_SIZES: () => PIE_CATEGORICAL_LEGEND_SIZES,
  PIE_INTERNAL_RADIUS: () => PIE_INTERNAL_RADIUS,
  SHOW_LABEL_RELATIVE_VALUE_THRESHOLD: () => SHOW_LABEL_RELATIVE_VALUE_THRESHOLD,
  SHOW_LABEL_SLICES_THRESHOLD: () => SHOW_LABEL_SLICES_THRESHOLD,
  THRESHOLD_TYPE: () => THRESHOLD_TYPE,
  defaultConfig: () => defaultConfig,
  zIndicesPieAreas: () => zIndicesPieAreas
});
module.exports = __toCommonJS(default_exports);
var import_spacings = __toESM(require("@dynatrace/strato-design-tokens/spacings"));
var import_constants = require("../../core/constants.js");
const DEFAULT_CONTAINER_PADDING = import_spacings.default.Size6;
const DEFAULT_SIZE = 300;
const DONUT_INTERNAL_RADIUS = 0.75;
const PIE_INTERNAL_RADIUS = 1;
const LABEL_FONT_SIZE = 12;
const SHOW_LABEL_RELATIVE_VALUE_THRESHOLD = 20;
const SHOW_LABEL_SLICES_THRESHOLD = 2;
const LABEL_FONT_SIZE_MULTIPLIER = 1.5;
const DEFAULT_LEGEND = {
  position: "auto",
  hidden: false,
  ratio: import_constants.defaultRatio
};
const DEFAULT_TOOLTIP_MARGIN = 40;
const defaultConfig = {
  height: DEFAULT_SIZE,
  colorPalette: "categorical",
  labels: { hidden: false, valueType: "relative" },
  grouping: { threshold: { type: "relative", value: 2 } },
  legend: DEFAULT_LEGEND,
  formatter: () => "",
  truncationMode: import_constants.DEFAULT_TRUNCATION_MODE
};
const zIndicesPieAreas = {
  toolbar: 2,
  overlay: 1
};
const THRESHOLD_TYPE = {
  ABSOLUTE: "absolute",
  RELATIVE: "relative",
  NUMBER_OF_SLICES: "number-of-slices"
};
const PIE_CATEGORICAL_LEGEND_SIZES = {
  minWidth: 0.1,
  minHeight: 0.1,
  maxWidth: 0.8,
  maxHeight: 0.7
};
const LABEL_TEXT_POSITION = {
  END: "end",
  START: "start"
};
