// @dynatrace/strato-components/esm/charts/categorical-bar/components/interactions/overlays/CategoricalBarChartInspectOverlay.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { useEffect, useState } from "react";
import {
  crosshairLayout,
  dataPointLayout
} from "./CategoricalBarChartInspectOverlay.sty.js";
import { plotLayout } from "../../../../core/components/chart-grid/ChartGrid.sty.js";
import { useGridPlotSize } from "../../../../core/components/chart-grid/hooks/useGridPlotSize.js";
import { HorizontalCrosshair } from "../../../../core/components/crosshair/HorizontalCrosshair.js";
import { getMagnifierLabel } from "../../../../core/components/crosshair/utils/get-magnifier-label.js";
import { VerticalCrosshair } from "../../../../core/components/crosshair/VerticalCrosshair.js";
import { THRESHOLD_BAR_THICKNESS } from "../../../../core/components/threshold-bar/types/constants.js";
import { useApplyCallbackOnScroll } from "../../../../core/hooks/useApplyCallbackOnScroll.js";
import {
  useMouseInBoundsState,
  useMousePositionState,
  useSetState,
  useTooltipState
} from "../../../hooks/store/selectors.js";
import { setMouseOutBounds } from "../../../hooks/store/state-setters.js";
import { useCategoricalBarChartBaseConfig } from "../../../hooks/useCategoricalBarChartBaseConfig.js";
import { useCategoricalBarChartData } from "../../../hooks/useCategoricalBarChartData.js";
import { useCategoricalBarChartOverlayRef } from "../../../hooks/useCategoricalBarChartOverlayRef.js";
import { useCategoricalBarChartScales } from "../../../hooks/useCategoricalBarChartScales.js";
import { useCategoricalBarChartThresholds } from "../../../hooks/useCategoricalBarChartThresholds.js";
import { useClosestDimensionInThresholdPoints } from "../../../hooks/useClosestDimensionInThresholdPoints.js";
import { useClosestDimensionPoint } from "../../../hooks/useClosestDimensionPoint.js";
import { useFormatter } from "../../../hooks/useFormatter.js";
import { isStacked, isVertical } from "../../../utils/layout.js";
import { CategoricalBarChartDatapoints } from "../datapoints/CategoricalBarChartDatapoints.js";
import { CategoricalBarChartTooltip } from "../tooltip/CategoricalBarChartTooltip.js";
import {
  generateGroupDataPoints,
  getGroupDataPointsBy,
  mapToDataPoint
} from "../utils/data-points.js";
const CROSSHAIR_BAR_CHART_LABEL = "crosshair for bar chart";
const CategoricalBarChartInspectOverlay = () => {
  const { layout, groupMode } = useCategoricalBarChartBaseConfig();
  const { width, height } = useGridPlotSize();
  const { unit } = useCategoricalBarChartData();
  const formatter = useFormatter();
  const { valueScale } = useCategoricalBarChartScales();
  const { visibleTooltip, pinnedTooltip } = useTooltipState();
  const mousePosition = useMousePositionState();
  const thresholds = useCategoricalBarChartThresholds();
  const chartRef = useCategoricalBarChartOverlayRef();
  const isInChartBounds = useMouseInBoundsState();
  const chartBounds = chartRef?.current?.getBoundingClientRect();
  const hasThresholds = thresholds && thresholds.length > 0;
  const [pinnedClosestPoint, setPinnedClosestPoint] = useState();
  const [pinnedGroupDataPoint, setPinnedGroupDataPoint] = useState();
  const magnifierCoordinate = isVertical(layout) ? "y" : "x";
  const magnifierValue = mousePosition && getMagnifierLabel(
    unit,
    magnifierCoordinate,
    formatter,
    mousePosition.relative,
    valueScale
  );
  const isGroupedAllowed = isStacked(groupMode);
  const { closestDimensionPoint, closestDimensionPoints } = useClosestDimensionPoint(layout, groupMode, mousePosition?.relative);
  const closestDimensionPointInThreshold = useClosestDimensionInThresholdPoints(
    layout,
    groupMode,
    mousePosition?.relative
  );
  const groupDataPoints = generateGroupDataPoints(
    closestDimensionPoints,
    isGroupedAllowed,
    layout,
    layout === "vertical" ? height : width
  );
  useEffect(() => {
    if (!visibleTooltip && !pinnedTooltip) {
      setPinnedClosestPoint(
        (prevState) => prevState !== void 0 ? void 0 : prevState
      );
      setPinnedGroupDataPoint(
        (prevState) => prevState !== void 0 ? void 0 : prevState
      );
    }
  }, [visibleTooltip, pinnedTooltip]);
  useEffect(() => {
    if (pinnedTooltip) {
      onClickHandler();
    }
  }, [width, height]);
  const setState = useSetState();
  const onScrollCallback = () => setState(setMouseOutBounds);
  const notShowingMagnifiers = pinnedTooltip || !isInChartBounds;
  useApplyCallbackOnScroll(notShowingMagnifiers, onScrollCallback);
  const onClickHandler = () => {
    setState((prev) => ({
      ...prev,
      overlays: {
        ...prev.overlays,
        pinnedTooltip: setPinned(prev.overlays.pinnedTooltip)
      }
    }));
    setPinnedClosestPoint(
      (prev) => !prev && visibleTooltip ? closestDimensionPointInThreshold ?? closestDimensionPoint : void 0
    );
    setPinnedGroupDataPoint((prev) => {
      if (!prev && closestDimensionPoint && visibleTooltip) {
        const dimensionId = (closestDimensionPointInThreshold ?? closestDimensionPoint)?.datapoint.id;
        return getGroupDataPointsBy(groupDataPoints, dimensionId);
      }
      return void 0;
    });
  };
  const setPinned = (prevPinned) => visibleTooltip ? !prevPinned : false;
  const verticalCrosshairX = pinnedTooltip && pinnedClosestPoint ? pinnedClosestPoint.x ?? 0 : mousePosition?.relative.x ?? 0;
  const verticalCrosshairLabel = pinnedTooltip || !mousePosition ? "" : magnifierValue;
  const horizontalCrosshairY = pinnedTooltip && pinnedClosestPoint ? pinnedClosestPoint.y ?? 0 : mousePosition?.relative.y ?? 0;
  const horizontalCrosshairLabel = pinnedTooltip || !mousePosition ? "" : magnifierValue;
  const closestPoint = pinnedTooltip && pinnedClosestPoint ? pinnedClosestPoint : closestDimensionPointInThreshold;
  const groupPoints = pinnedGroupDataPoint ? pinnedGroupDataPoint : groupDataPoints;
  const thresholdSize = hasThresholds ? THRESHOLD_BAR_THICKNESS : 0;
  return /* @__PURE__ */ jsxs("div", { className: plotLayout, children: [
    (mousePosition || pinnedTooltip) && /* @__PURE__ */ jsxs(Fragment, { children: [
      /* @__PURE__ */ jsx("div", { className: dataPointLayout, children: /* @__PURE__ */ jsx(
        CategoricalBarChartDatapoints,
        {
          points: groupPoints,
          closestPoint: closestPoint && mapToDataPoint(closestPoint)
        }
      ) }),
      /* @__PURE__ */ jsx("div", { children: /* @__PURE__ */ jsx(
        CategoricalBarChartTooltip,
        {
          closestDimensionPoint: closestPoint,
          hasHeader: groupPoints.length > 1,
          formatter,
          unit
        }
      ) })
    ] }),
    pinnedTooltip || isInChartBounds ? /* @__PURE__ */ jsx(
      "svg",
      {
        className: crosshairLayout,
        "aria-label": CROSSHAIR_BAR_CHART_LABEL,
        role: "graphics-symbol",
        onClick: onClickHandler,
        children: !isVertical(layout) ? /* @__PURE__ */ jsx(
          VerticalCrosshair,
          {
            x: verticalCrosshairX,
            bottomLabel: verticalCrosshairLabel,
            height: height + thresholdSize,
            offset: thresholdSize,
            dashed: !pinnedTooltip,
            chartBounds
          }
        ) : /* @__PURE__ */ jsx(
          HorizontalCrosshair,
          {
            width: width + thresholdSize,
            leftLabel: horizontalCrosshairLabel,
            leftOffset: thresholdSize,
            y: horizontalCrosshairY,
            dashed: !pinnedTooltip,
            chartBounds
          }
        )
      }
    ) : null
  ] });
};
CategoricalBarChartInspectOverlay["displayName"] = "CategoricalBarChartInspectOverlay";
export {
  CROSSHAIR_BAR_CHART_LABEL,
  CategoricalBarChartInspectOverlay
};
