// @dynatrace/strato-components/filters/filter-field/utils/conversion/lezer-tree-conversion/value/free-text.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var free_text_exports = {};
__export(free_text_exports, {
  convertFreeText: () => convertFreeText
});
module.exports = __toCommonJS(free_text_exports);
var import_decoration_marks = require("../../../../decorations/decoration-marks.js");
var import_get_text = require("../../../get-text.js");
var import_validation_helper_functions = require("../../../validation-utils/validation-helper-functions.js");
var import_escape_characters = require("../../escape-characters.js");
function convertFreeText(node, doc, parserConfig) {
  const { from, to } = node;
  const text = (0, import_get_text.getTextFromRange)(node, doc);
  const value = (0, import_escape_characters.removeEscapeCharacters)(text);
  const errors = [];
  const convertedNode = {
    type: "Error",
    range: {
      from,
      to
    },
    textValue: text,
    value
  };
  if (parserConfig?.searchConversion) {
    errors.push({
      range: { from: to - 1, to },
      message: {
        type: "incompleteStatementWithEnabledSearchError",
        details: (0, import_validation_helper_functions.getColoredSearchOperator)()
      },
      decoration: import_decoration_marks.postSpotErrorMark
    });
  } else {
    errors.push({
      range: { from: to - 1, to },
      message: "incompleteStatementError",
      decoration: import_decoration_marks.postSpotErrorMark
    });
  }
  return {
    convertedNode,
    errors
  };
}
