// @dynatrace/strato-components/filters/filter-field/utils/conversion/lezer-tree-conversion/value/identifier.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var identifier_exports = {};
__export(identifier_exports, {
  convertIdentifier: () => convertIdentifier
});
module.exports = __toCommonJS(identifier_exports);
var import_get_text = require("../../../get-text.js");
var import_escape_characters = require("../../escape-characters.js");
function convertIdentifier(node, doc, context) {
  const { from, to } = node;
  const text = (0, import_get_text.getTextFromRange)(node, doc);
  const value = (0, import_escape_characters.removeEscapeCharacters)(
    context === "String" ? text.slice(1, -1) : text
  );
  const convertedNode = {
    type: "String",
    range: {
      from,
      to
    },
    textValue: text,
    value
  };
  let endsWithOperator;
  let startsWithOperator;
  if (node.name === "String") {
    endsWithOperator = text.slice(1, 2) === "*" ? { from: from + 1, to: from + 2 } : void 0;
    const withoutQuotes = text.slice(1, text.length - 1);
    startsWithOperator = withoutQuotes.match(/(?<!\\)(?:\\\\)*[*]$/) ? { from: to - 2, to: to - 1 } : void 0;
    convertedNode.value = value.substring(
      endsWithOperator ? 1 : 0,
      value.length - (startsWithOperator ? 1 : 0)
    );
  }
  return {
    convertedNode,
    errors: [],
    isValid: (0, import_escape_characters.isValidEscapeSequence)(text),
    positionOperators: {
      endsWithOperator,
      startsWithOperator
    }
  };
}
