// @dynatrace/strato-components/filters/filter-field/utils/conversion/lezer-tree-conversion/value/simple-value.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var simple_value_exports = {};
__export(simple_value_exports, {
  convertSimpleValue: () => convertSimpleValue
});
module.exports = __toCommonJS(simple_value_exports);
var import_is_node_environment = require("../../../../../../core/utils/is-node-environment.js");
var import_get_text = require("../../../get-text.js");
var import_type_validators = require("../../../validation-utils/type-validators.js");
var import_generic_node = require("../generic-node.js");
var import_node_mutations = require("../node-mutations.js");
var import_boolean = require("./boolean.js");
var import_duration = require("./duration.js");
var import_identifier = require("./identifier.js");
var import_number = require("./number.js");
var import_validation_helper_functions = require("../../../validation-utils/validation-helper-functions.js");
function convertSimpleValue(node, doc, numberParser, customTypes, parserConfig) {
  const { from, to } = node;
  const cursor = node.node.cursor();
  let convertedNode = {
    type: "String",
    range: {
      from,
      to
    },
    textValue: "",
    value: ""
    // With the type defaulting to 'String', typescript would infer a 'FilterFieldStringNode'.
  };
  const errors = [];
  let positionOperators;
  let positionOperatorsInsideString;
  cursor.iterate((childNode) => {
    const { from: childFrom, to: childTo } = childNode;
    const nodeName = childNode.name;
    switch (nodeName) {
      case "FilterSimpleValue":
      case "ListValue":
        break;
      case "Identifier":
      case "String":
      case "IncompleteString":
      case "Variable": {
        const {
          convertedNode: { textValue, value },
          positionOperators: { endsWithOperator, startsWithOperator },
          isValid: isValidIdentifier
        } = (0, import_identifier.convertIdentifier)(childNode, doc, nodeName);
        if (
          // eslint-disable-next-line @typescript-eslint/no-deprecated
          parserConfig?.ipAddressConversion && nodeName === "Identifier" && (0, import_type_validators.isValidIPAddress)(value)
        ) {
          convertedNode.type = "IPAddress";
        } else if (
          // eslint-disable-next-line @typescript-eslint/no-deprecated
          parserConfig?.timestampConversion && (nodeName === "Identifier" || nodeName === "String") && (0, import_type_validators.isValidTimestamp)(value)
        ) {
          convertedNode.type = "Timestamp";
        }
        if (
          // eslint-disable-next-line @typescript-eslint/no-deprecated
          parserConfig?.uidConversion && nodeName === "Identifier" && (0, import_type_validators.isValidUID)(value)
        ) {
          convertedNode.type = "UID";
        }
        if (
          // eslint-disable-next-line @typescript-eslint/no-deprecated
          parserConfig?.smartscapeIdConversion && nodeName === "Identifier" && (0, import_type_validators.isValidSmartscapeId)(value)
        ) {
          convertedNode.type = "SmartscapeId";
        }
        const customType = (0, import_validation_helper_functions.isValidCustomType)(value, customTypes);
        if (customType) {
          convertedNode.type = customType;
        }
        if (endsWithOperator || startsWithOperator) {
          if (!positionOperatorsInsideString) {
            positionOperatorsInsideString = { ranges: [] };
          }
          if (endsWithOperator) {
            positionOperatorsInsideString.type = "ends-with";
            positionOperatorsInsideString.ranges.push(endsWithOperator);
          }
          if (startsWithOperator) {
            positionOperatorsInsideString.type = "starts-with";
            positionOperatorsInsideString.ranges.push(startsWithOperator);
          }
          if (endsWithOperator && startsWithOperator) {
            positionOperatorsInsideString.type = "contains";
            if (endsWithOperator.from === startsWithOperator.from && endsWithOperator.to === startsWithOperator.to) {
              (0, import_node_mutations.convertToErrorNode)(convertedNode);
              errors.push({
                range: positionOperatorsInsideString.ranges[0],
                message: "unsupportedPositionOperator"
              });
            }
          }
        }
        (0, import_node_mutations.updateNodeValues)(convertedNode, textValue, value);
        if (nodeName === "Variable") {
          convertedNode.type = "Variable";
          convertedNode.variableType = "Any";
        }
        if (!isValidIdentifier || nodeName === "IncompleteString") {
          (0, import_node_mutations.convertToErrorNode)(convertedNode);
        }
        break;
      }
      case "InvalidIdentifier": {
        const {
          convertedNode: { textValue }
        } = (0, import_generic_node.convertNode)(childNode, doc);
        (0, import_node_mutations.updateNodeValues)(convertedNode, textValue);
        (0, import_node_mutations.convertToErrorNode)(convertedNode);
        break;
      }
      case "Boolean": {
        const { convertedNode: booleanNode } = (0, import_boolean.convertBoolean)(childNode, doc);
        convertedNode = {
          ...convertedNode,
          ...booleanNode
        };
        break;
      }
      case "Number": {
        ({ convertedNode } = (0, import_number.convertNumber)(childNode, doc, numberParser));
        break;
      }
      case "Duration":
        ({ convertedNode } = (0, import_duration.convertDuration)(childNode, doc));
        break;
      case "PositionOperator": {
        const text = (0, import_get_text.getTextFromRange)(childNode, doc);
        (0, import_node_mutations.updateNodeValues)(convertedNode, text, "");
        if (!positionOperators) {
          positionOperators = { ranges: [] };
        }
        const childRange = { from: childFrom, to: childTo };
        if (childFrom === from) {
          positionOperators.type = "ends-with";
          positionOperators.ranges.push(childRange);
        } else if (childTo === to) {
          positionOperators.type = positionOperators.type ? "contains" : "starts-with";
          positionOperators.ranges.push(childRange);
        }
        break;
      }
      case "InvalidPositionOperator": {
        const text = (0, import_get_text.getTextFromRange)(childNode, doc);
        (0, import_node_mutations.updateNodeValues)(convertedNode, text);
        (0, import_node_mutations.convertToErrorNode)(convertedNode);
        if (nodeName === "InvalidPositionOperator") {
          errors.push({
            range: { from: childFrom, to: childTo },
            message: "unsupportedPositionOperator"
          });
        }
        break;
      }
      case "\u26A0": {
        convertedNode.value = convertedNode.textValue;
        (0, import_node_mutations.convertToErrorNode)(convertedNode);
        return false;
      }
      default:
        if ((0, import_is_node_environment.isDev)()) {
          console.log(
            `${doc.toString()} | Syntax node '${childNode.name}' is not yet handled in the simple value conversion. Please make sure to handle all nodes accordingly.`
          );
        }
    }
  });
  if (convertedNode.type === "Error") {
    convertedNode.textValue = convertedNode.value = (0, import_get_text.getTextFromRange)(
      convertedNode.range,
      doc
    );
    errors.push({
      range: { from, to },
      message: "invalidValueErrorGeneric"
    });
  }
  let combinedPositionOperators = positionOperators ?? positionOperatorsInsideString;
  if (positionOperators && positionOperatorsInsideString) {
    combinedPositionOperators = {
      ...positionOperators,
      ranges: [
        ...positionOperators?.ranges ?? [],
        ...positionOperatorsInsideString?.ranges ?? []
      ]
    };
    const { type } = positionOperators;
    const { type: typeInsideString } = positionOperatorsInsideString;
    if (
      // foo = *"*bar" | foo = *"*bar"*
      typeInsideString === "ends-with" && (type === "ends-with" || type === "contains") || // foo = "bar*"* | foo = *"bar*"*
      typeInsideString === "starts-with" && (type === "starts-with" || type === "contains") || // foo = *"*bar*"*
      typeInsideString === "contains" && type === "contains"
    ) {
      (0, import_node_mutations.convertToErrorNode)(convertedNode);
      errors.push({
        range: positionOperatorsInsideString.ranges[0],
        message: "duplicatePositionOperator"
      });
      errors.push({
        range: positionOperators.ranges[0],
        message: "duplicatePositionOperator"
      });
    }
    if (typeInsideString === "contains" && type === "contains") {
      (0, import_node_mutations.convertToErrorNode)(convertedNode);
      errors.push({
        range: positionOperatorsInsideString.ranges[1],
        message: "duplicatePositionOperator"
      });
      errors.push({
        range: positionOperators.ranges[1],
        message: "duplicatePositionOperator"
      });
    }
    if (typeInsideString === "ends-with" && type === "starts-with" || typeInsideString === "starts-with" && type === "ends-with") {
      combinedPositionOperators.type = "contains";
    }
  }
  return {
    convertedNode,
    errors,
    positionOperators: errors.length === 0 ? combinedPositionOperators : void 0
  };
}
