// @dynatrace/strato-components/charts/tree-map/components/canvas/utils/canvas-label-utils.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var canvas_label_utils_exports = {};
__export(canvas_label_utils_exports, {
  FONT_SIZE_DIFFERENCE: () => FONT_SIZE_DIFFERENCE,
  LINE_SPACING: () => LINE_SPACING,
  getAppliedLabels: () => getAppliedLabels,
  getLabelFit: () => getLabelFit,
  getLabelWidth: () => getLabelWidth,
  getOffsetY: () => getOffsetY,
  getRectCenter: () => getRectCenter
});
module.exports = __toCommonJS(canvas_label_utils_exports);
var import_truncate_string = require("../../../../core/components/text-ellipsis/truncate-string.js");
var import_constants = require("../../../constants.js");
const LINE_SPACING = 2;
const FONT_SIZE_DIFFERENCE = 1;
const getLabelFit = (height, width, fontSize, padding, labelWidth) => {
  const { nameWidth, nameValueWidth, valueWidth, multiLineWidth } = labelWidth;
  const lineSpacing = 2;
  const paddingSum = 2 * padding;
  const availableWidth = width - paddingSum;
  const minInlineHeight = paddingSum + fontSize;
  const minMultiLineHeight = paddingSum + fontSize * 2 + lineSpacing;
  const fitMultiLine = height >= minMultiLineHeight && availableWidth >= multiLineWidth;
  if (fitMultiLine) {
    return import_constants.LABEL_OPTIONS.Multi;
  }
  const fitInline = height >= minInlineHeight && availableWidth >= nameValueWidth;
  if (fitInline) {
    return import_constants.LABEL_OPTIONS.Inline;
  }
  const fitName = height >= minInlineHeight && availableWidth >= nameWidth;
  if (fitName) {
    return import_constants.LABEL_OPTIONS.Name;
  }
  const fitEllipsis = height >= minInlineHeight && availableWidth < nameWidth;
  if (fitEllipsis) {
    const fitAtLeastValue = height >= minMultiLineHeight && availableWidth >= valueWidth;
    return fitAtLeastValue ? import_constants.LABEL_OPTIONS.EllipsisWithValue : import_constants.LABEL_OPTIONS.Ellipsis;
  }
  return import_constants.LABEL_OPTIONS.Fallback;
};
const getRectCenter = (width, height) => ({
  centerX: width / 2,
  centerY: height / 2
});
const getLabelWidth = (nameWidth, value, charWidthMap) => {
  const valueWidth = (0, import_truncate_string.calculateWidth)(value, charWidthMap);
  const nameValueWidth = nameWidth + (0, import_truncate_string.calculateWidth)(` ${value}`, charWidthMap);
  const multiLineWidth = Math.max(nameWidth, valueWidth);
  return { nameWidth, valueWidth, nameValueWidth, multiLineWidth };
};
const getAppliedLabels = (labelFit, name, value) => {
  const labelMap = {
    multi: [name, value],
    inline: [`${name} ${value}`],
    name: [name],
    ellipsis: [name],
    "ellipsis-with-value": [name, value],
    fallback: []
  };
  return labelMap[labelFit];
};
const getOffsetY = (labels, labelIndex, fontSize) => {
  const downOffsetY = Math.round(fontSize / 2) + LINE_SPACING;
  const upOffsetY = -1 * downOffsetY - FONT_SIZE_DIFFERENCE;
  const isMultiLine = labels.length > 1;
  return isMultiLine ? labelIndex === 0 ? upOffsetY : downOffsetY : 0;
};
