// @dynatrace/strato-components/charts/gauge/components/GaugeChartTooltip/GaugeOverlayTooltip.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var GaugeOverlayTooltip_exports = {};
__export(GaugeOverlayTooltip_exports, {
  GaugeOverlayTooltip: () => GaugeOverlayTooltip
});
module.exports = __toCommonJS(GaugeOverlayTooltip_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_intl = require("react-intl");
var import_OverlayTooltip = require("../../../core/components/overlay-tooltip/OverlayTooltip.js");
var import_useTooltipSeriesActions = require("../../../core/components/tooltip/shared/hooks/useTooltipSeriesActions.js");
var import_GaugeChartTooltipContext = require("../../contexts/GaugeChartTooltipContext.js");
var import_useGaugeChartFormatterContext = require("../../hooks/useGaugeChartFormatterContext.js");
const GaugeOverlayTooltip = ({
  value,
  color
}) => {
  const intl = (0, import_react_intl.useIntl)();
  const { formatter, unit } = (0, import_useGaugeChartFormatterContext.useGaugeChartFormatterContext)();
  const { name: tooltipName, template } = (0, import_GaugeChartTooltipContext.useGaugeChartTooltip)();
  const seriesName = tooltipName ?? intl.formatMessage({ defaultMessage: "Value", id: "jc4XDeI4sf7aLs2W" });
  const formattedValue = formatter(value, unit);
  const tooltipItem = {
    id: `${seriesName}-${value}`,
    name: seriesName,
    values: [formattedValue]
  };
  const seriesActions = (0, import_useTooltipSeriesActions.useTooltipSeriesActions)(
    [tooltipItem],
    (item) => item.id
  );
  const actions = tooltipItem ? () => seriesActions(tooltipItem) : void 0;
  if (template) {
    const customItem = {
      value,
      formattedValue,
      color,
      name: tooltipName,
      unit: typeof unit === "string" ? unit : void 0
    };
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayTooltip.OverlayTooltip, { legacyTemplate: template(customItem) });
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayTooltip.OverlayTooltip, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_OverlayTooltip.OverlayTooltip.SeriesItem,
    {
      name: seriesName,
      value: formattedValue,
      color,
      actions
    }
  ) });
};
GaugeOverlayTooltip["displayName"] = "GaugeOverlayTooltip";
