// @dynatrace/strato-components/charts/histogram/components/overlays/utils/bins-to-tooltip.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var bins_to_tooltip_exports = {};
__export(bins_to_tooltip_exports, {
  toTooltipItems: () => toTooltipItems
});
module.exports = __toCommonJS(bins_to_tooltip_exports);
var import_get_formatted_values = require("./get-formatted-values.js");
var import_HistogramDataProvider = require("../../../providers/HistogramDataProvider.js");
function toTooltipItems(binsMap, formatters, isSingle, units, points, closestPoint) {
  if (!closestPoint) {
    return { tooltipItems: [], closestTooltipItem: void 0 };
  }
  const binMetadata = (0, import_HistogramDataProvider.safeGet)(binsMap, closestPoint.bin);
  const [leftUnit] = units;
  const [leftFormatter, rightFormatter] = formatters;
  const formatter = binMetadata.unit === leftUnit ? leftFormatter : rightFormatter;
  const closestTooltipItem = {
    id: binMetadata.seriesId,
    name: binMetadata.seriesName,
    color: binMetadata.color,
    asyncValues: () => (0, import_get_formatted_values.getFormattedValues)(formatter, binMetadata),
    bin: closestPoint.bin
  };
  if (isSingle) {
    return {
      tooltipItems: [closestTooltipItem],
      closestTooltipItem
    };
  }
  if (!points) {
    return { tooltipItems: [], closestTooltipItem };
  }
  const tooltipItems = points.sort((a, b) => a.y - b.y).map((point) => {
    const binSharedMetadata = (0, import_HistogramDataProvider.safeGet)(binsMap, point.bin);
    return {
      id: binSharedMetadata.seriesId,
      name: binSharedMetadata.seriesName,
      color: binSharedMetadata.color,
      asyncValues: () => (0, import_get_formatted_values.getFormattedValues)(formatter, binSharedMetadata),
      bin: point.bin
    };
  });
  return { tooltipItems, closestTooltipItem };
}
