// @dynatrace/strato-components/esm/charts/single-value/utils/single-value-config.js
import { merge } from "lodash-es";
import { isValidElement } from "react";
import { getToolbarConfigWithoutIntents } from "../../core/components/toolbar/utils/get-toolbar-config-without-intents.js";
import { isValidColorRule } from "../../core/slots/ColorRule.js";
import { defaultSingleValueConfig } from "../defaults.js";
import {
  defaultToolbarInternalConfig
} from "../types/single-value-toolbar.js";
function getNodeInnerText(node) {
  if (typeof node === "string") {
    return node;
  }
  if (node === null || typeof node === "boolean" || typeof node === "undefined") {
    return "";
  }
  if (typeof node === "number") {
    return node.toString();
  }
  if (isValidElement(node) && node.props && node.props.children) {
    return getNodeInnerText(node.props.children);
  }
  return "";
}
function deserializeConfig(serializedConfig) {
  try {
    return merge({}, defaultSingleValueConfig, JSON.parse(serializedConfig));
  } catch {
    return defaultSingleValueConfig;
  }
}
function isSerializedConfig(contextConfig) {
  return typeof contextConfig === "string";
}
function normalizeConfig(config) {
  const sparklineConfig = config.sparkline;
  const toolbar = config.toolbar ? getToolbarConfigWithoutIntents(config.toolbar) : void 0;
  const output = {
    ...config,
    label: getNodeInnerText(config.label),
    toolbar
  };
  if (sparklineConfig) {
    output.sparkline = sparklineConfig;
  }
  return output;
}
const buildSparklineConfig = (slotsSparklineConfig, contextSparklineConfig) => {
  if (!slotsSparklineConfig) {
    return;
  }
  const contextThresholds = contextSparklineConfig?.thresholdIndicators || [];
  const slotThresholds = slotsSparklineConfig?.thresholdIndicators || [];
  return {
    thresholdIndicators: contextThresholds.concat(slotThresholds),
    color: slotsSparklineConfig?.color ?? contextSparklineConfig?.color,
    variant: slotsSparklineConfig?.variant ?? contextSparklineConfig?.variant,
    xAxis: {
      ...contextSparklineConfig?.xAxis,
      ...slotsSparklineConfig?.xAxis
    },
    yAxis: {
      ...contextSparklineConfig?.yAxis,
      ...slotsSparklineConfig?.yAxis
    },
    showTicks: slotsSparklineConfig?.showTicks ?? contextSparklineConfig?.showTicks,
    showContextValues: slotsSparklineConfig?.showContextValues ?? contextSparklineConfig?.showContextValues,
    gapPolicy: slotsSparklineConfig?.gapPolicy ?? contextSparklineConfig?.gapPolicy,
    curve: slotsSparklineConfig?.curve
  };
};
const getDefaultToolbarPlacement = (singleValueAlignment) => {
  return singleValueAlignment === "end" ? "top-left" : defaultToolbarInternalConfig.placement;
};
const getToolbarConfig = (toolbarSlotsConfig, toolbarConsumerConfig, singleValueAlignment) => {
  if (!toolbarSlotsConfig && !toolbarConsumerConfig) {
    return void 0;
  }
  const placement = toolbarSlotsConfig?.placement ?? toolbarConsumerConfig?.placement ?? getDefaultToolbarPlacement(singleValueAlignment);
  return merge(
    {},
    defaultToolbarInternalConfig,
    toolbarConsumerConfig,
    toolbarSlotsConfig,
    { placement }
  );
};
function buildSingleValueConfig(props, slots, contextConfig) {
  const colorRules = (contextConfig.colorRules || []).filter(isValidColorRule).concat(slots.colorRules || []);
  const toolbar = getToolbarConfig(
    slots.toolbar,
    contextConfig.toolbar,
    props.alignment
  );
  return {
    label: props.label ?? contextConfig.label,
    color: props.color ?? contextConfig.color,
    unit: props.unit ?? contextConfig.unit ?? defaultSingleValueConfig.unit,
    formatter: props.formatter ?? contextConfig.formatter,
    alignment: props.alignment ?? contextConfig.alignment ?? defaultSingleValueConfig.alignment,
    sparkline: buildSparklineConfig(slots.sparkline, contextConfig.sparkline),
    colorRules,
    toolbar
  };
}
export {
  buildSingleValueConfig,
  buildSparklineConfig,
  deserializeConfig,
  isSerializedConfig,
  normalizeConfig
};
