// @dynatrace/strato-design-tokens/animations-js/index.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var animations_js_exports = {};
__export(animations_js_exports, {
  default: () => animations_js_default
});
module.exports = __toCommonJS(animations_js_exports);
const Open = {
  Duration: {
    /** Value: 0.17. */
    Fast: 0.17,
    /** Value: 0.25. */
    Base: 0.25,
    /** Value: 0.4. */
    Slow: 0.4
  },
  /** Value: [ 0, 0,0.2,1 ]. */
  Easing: [0, 0, 0.2, 1]
};
const Close = {
  Duration: {
    /** Value: 0. */
    Fast: 0,
    /** Value: 0.17. */
    Base: 0.17,
    /** Value: 0.25. */
    Slow: 0.25
  },
  /** Value: [ 0.5, 0,1,1 ]. */
  Easing: [0.5, 0, 1, 1]
};
const SlideIn = {
  Duration: {
    /** Value: 0.2. */
    Fast: 0.2,
    /** Value: 0.3. */
    Base: 0.3,
    /** Value: 0.4. */
    Slow: 0.4
  },
  /** Value: [ 0, 0,0.2,1 ]. */
  Easing: [0, 0, 0.2, 1]
};
const SlideOut = {
  Duration: {
    /** Value: 0.15. */
    Fast: 0.15,
    /** Value: 0.25. */
    Base: 0.25,
    /** Value: 0.35. */
    Slow: 0.35
  },
  /** Value: [ 0.8, 0,0.5,1 ]. */
  Easing: [0.8, 0, 0.5, 1]
};
const FadeIn = {
  Duration: {
    /** Value: 0.15. */
    Fast: 0.15,
    /** Value: 0.25. */
    Base: 0.25,
    /** Value: 0.35. */
    Slow: 0.35
  },
  /** Value: [ 0.5, 0,1,1 ]. */
  Easing: [0.5, 0, 1, 1]
};
const FadeOut = {
  Duration: {
    /** Value: 0.15. */
    Fast: 0.15,
    /** Value: 0.25. */
    Base: 0.25,
    /** Value: 0.35. */
    Slow: 0.35
  },
  /** Value: [ 0, 0,0.2,1 ]. */
  Easing: [0, 0, 0.2, 1]
};
const Hover = {
  Duration: {
    /** Value: 0.15. */
    Base: 0.15
  },
  /** Value: [ 0, 0,0.2,1 ]. */
  Easing: [0, 0, 0.2, 1]
};
const StaggeredMotion = {
  Duration: {
    /** Value: 0.25. */
    Fast: 0.25,
    /** Value: 0.35. */
    Base: 0.35,
    /** Value: 0.4. */
    Slow: 0.4
  },
  Offset: {
    /** Value: 0.02. */
    Fast: 0.02,
    /** Value: 0.03. */
    Base: 0.03,
    /** Value: 0.05. */
    Slow: 0.05
  },
  /** Value: [ 0, 0,0.2,1 ]. */
  Easing: [0, 0, 0.2, 1]
};
const ExpandCollapse = {
  Duration: {
    /** Value: 0.25. */
    Fast: 0.25,
    /** Value: 0.4. */
    Base: 0.4,
    /** Value: 0.6. */
    Slow: 0.6
  },
  /** Value: [ 0.4, 0,0.2,1 ]. */
  Easing: [0.4, 0, 0.2, 1]
};
const Move = {
  Duration: {
    /** Value: 0.15. */
    Fast: 0.15,
    /** Value: 0.4. */
    Base: 0.4,
    /** Value: 0.6. */
    Slow: 0.6
  },
  /** Value: [ 0.4, 0,0.2,1 ]. */
  Easing: [0.4, 0, 0.2, 1]
};
const Rescale = {
  Duration: {
    /** Value: 0.17. */
    Fast: 0.17,
    /** Value: 0.25. */
    Base: 0.25,
    /** Value: 0.4. */
    Slow: 0.4
  },
  /** Value: [ 0.4, 0,0.2,1 ]. */
  Easing: [0.4, 0, 0.2, 1]
};
var animations_js_default = {
  Open,
  Close,
  SlideIn,
  SlideOut,
  FadeIn,
  FadeOut,
  Hover,
  StaggeredMotion,
  ExpandCollapse,
  Move,
  Rescale
};
