// @dynatrace/strato-components/charts/timeseries/components/axes/TimeseriesValueLeftAxisArea.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TimeseriesValueLeftAxisArea_exports = {};
__export(TimeseriesValueLeftAxisArea_exports, {
  TimeseriesValueLeftAxisArea: () => TimeseriesValueLeftAxisArea
});
module.exports = __toCommonJS(TimeseriesValueLeftAxisArea_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_ValueLeftAxis = require("../../../core/components/axes/components/ValueLeftAxis.js");
var import_ChartGridRoot = require("../../../core/components/chart-grid/ChartGridRoot.js");
var import_useTimeseriesFilteredYAxis = require("../../hooks/useTimeseriesFilteredYAxis.js");
var import_useTimeseriesTickAxis = require("../../hooks/useTimeseriesTickAxis.js");
var import_useYScales = require("../../hooks/useYScales.js");
var import_timeseries_grid_area_rules = require("../../types/timeseries-grid-area-rules.js");
const TimeseriesValueLeftAxisArea = () => {
  const { left: axisLeftConfig } = (0, import_useTimeseriesFilteredYAxis.useTimeseriesFilteredYAxis)();
  const { left: scaleLeft } = (0, import_useYScales.useYScales)();
  const tickAxis = (0, import_useTimeseriesTickAxis.useTimeseriesTickAxis)();
  const leftTickAxis = tickAxis?.left || { ticks: [], formattedTicks: [] };
  if (!leftTickAxis.ticks.length || !scaleLeft) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ChartGridRoot.ChartGrid.GridArea, { gridArea: import_timeseries_grid_area_rules.AREAS.leftAxis, children: ({ availableWidth, availableHeight }) => {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_ValueLeftAxis.ValueLeftAxis,
      {
        ticks: leftTickAxis.ticks,
        formattedTicks: leftTickAxis.formattedTicks,
        containerHeight: availableHeight,
        containerWidth: availableWidth,
        scale: scaleLeft.scale,
        label: axisLeftConfig?.label
      }
    );
  } });
};
TimeseriesValueLeftAxisArea["displayName"] = "TimeseriesValueLeftAxisArea";
