// @dynatrace/strato-components/charts/timeseries/components/axes/TimeseriesValueRightAxisArea.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TimeseriesValueRightAxisArea_exports = {};
__export(TimeseriesValueRightAxisArea_exports, {
  TimeseriesValueRightAxisArea: () => TimeseriesValueRightAxisArea
});
module.exports = __toCommonJS(TimeseriesValueRightAxisArea_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_ValueRightAxis = require("../../../core/components/axes/components/ValueRightAxis.js");
var import_ChartGridRoot = require("../../../core/components/chart-grid/ChartGridRoot.js");
var import_useTimeseriesFilteredYAxis = require("../../hooks/useTimeseriesFilteredYAxis.js");
var import_useTimeseriesTickAxis = require("../../hooks/useTimeseriesTickAxis.js");
var import_useYScales = require("../../hooks/useYScales.js");
var import_timeseries_grid_area_rules = require("../../types/timeseries-grid-area-rules.js");
const TimeseriesValueRightAxisArea = () => {
  const { right: rightConfig } = (0, import_useTimeseriesFilteredYAxis.useTimeseriesFilteredYAxis)();
  const { right: scaleRight } = (0, import_useYScales.useYScales)();
  const tickAxis = (0, import_useTimeseriesTickAxis.useTimeseriesTickAxis)();
  const rightTickAxis = tickAxis?.right || { ticks: [], formattedTicks: [] };
  if (!rightTickAxis.ticks.length || !scaleRight) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ChartGridRoot.ChartGrid.GridArea, { gridArea: import_timeseries_grid_area_rules.AREAS.rightAxis, children: ({ availableWidth, availableHeight }) => {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_ValueRightAxis.ValueRightAxis,
      {
        ticks: rightTickAxis.ticks,
        formattedTicks: rightTickAxis.formattedTicks,
        containerHeight: availableHeight,
        containerWidth: availableWidth,
        scale: scaleRight.scale,
        label: rightConfig?.label
      }
    );
  } });
};
TimeseriesValueRightAxisArea["displayName"] = "TimeseriesValueRightAxisArea";
