// @dynatrace/strato-components/esm/charts/core/components/tooltip/shared/styles/SeriesBasedContent.sty.js
import "../../../../../../styles/textStyle.css";
import "./SeriesBasedContent.css";
import { createRuntimeFn as _7a468 } from "@vanilla-extract/recipes/createRuntimeFn";
var detailViewContent = "_dacoys0-3-6-0";
var detailViewItem = "_dacoys1-3-6-0";
var multipleDimensionContainer = _7a468({ defaultClassName: "_dacoys5-3-6-0", variantClassNames: { highlighted: { true: "_dacoys6-3-6-0", false: "_dacoys7-3-6-0" } }, defaultVariants: {}, compoundVariants: [] });
var multipleDimensionItem = "_dacoys8-3-6-0";
var singleDimensionContainer = _7a468({ defaultClassName: "_dacoys2-3-6-0", variantClassNames: { highlighted: { true: "_dacoys3-3-6-0", false: "_dacoys4-3-6-0" } }, defaultVariants: {}, compoundVariants: [] });
var valueContainer = "_dacoys9-3-6-0";
var valueItem = "_dacoysa-3-6-0 _da9a8v0-3-6-0 _da9a8v2-3-6-0 _da9a8v6-3-6-0 _da9a8ve-3-6-0";
export {
  detailViewContent,
  detailViewItem,
  multipleDimensionContainer,
  multipleDimensionItem,
  singleDimensionContainer,
  valueContainer,
  valueItem
};
