// @dynatrace/strato-components/editors/base-code-editor/BaseCodeEditor.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var BaseCodeEditor_exports = {};
__export(BaseCodeEditor_exports, {
  BaseCodeEditor: () => BaseCodeEditor
});
module.exports = __toCommonJS(BaseCodeEditor_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_autocomplete = require("@codemirror/autocomplete");
var import_language = require("@codemirror/language");
var import_state = require("@codemirror/state");
var import_view = require("@codemirror/view");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_intl = require("react-intl");
var styles = __toESM(require("./BaseCodeEditor.sty.js"));
var import_configure_line_wrap = require("./configuration/configure-line-wrap.js");
var import_configure_placeholder = require("./configuration/configure-placeholder.js");
var import_configure_read_only = require("./configuration/configure-read-only.js");
var import_configure_spell_check = require("./configuration/configure-spell-check.js");
var import_getDefaultConfigurations = require("./configuration/getDefaultConfigurations.js");
var import_language_extensions = require("./configuration/language-extensions.js");
var import_highlight_extension = require("./highlight/highlight.extension.js");
var import_testIds = require("./testIds.js");
var import_calculate_folding = require("./utils/calculate-folding.js");
var import_key_maps = require("./utils/key-maps.js");
var import_run_foldings = require("./utils/run-foldings.js");
var import_sanitize_text = require("./utils/sanitize-text.js");
var import_sanitize_value = require("./utils/sanitize-value.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useControllableState = require("../../core/hooks/useControllableState.js");
var import_useFormControlState = require("../../core/hooks/useFormControlState.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_useFocusRing = require("../../core/styles/useFocusRing.js");
var import_FormFieldMessageBaseItem = require("../../forms/forms-core/components/form-field-messages/FormFieldMessageBaseItem.js");
var import_useSyncedFormControlProps = require("../../forms/forms-core/hooks/useSyncedFormControlProps.js");
var import_InputGroup_css = require("../../layouts/input-group/InputGroup.sty.js");
import_view.EditorView.EDIT_CONTEXT = false;
const languageCompartment = new import_state.Compartment();
const editorBaseConfigCompartment = new import_state.Compartment();
const lineWrappingCompartment = new import_state.Compartment();
const keymapCompartment = new import_state.Compartment();
const placeholderCompartment = new import_state.Compartment();
const highlightOnChangeCompartment = new import_state.Compartment();
const getEditorConfiguration = (intlProps, options) => {
  const { singleLine, gutterConfiguration, hideLineNumbers, required, size } = options;
  const isRequired = required ?? false;
  const editorSize = size ?? "default";
  if (singleLine) {
    return (0, import_getDefaultConfigurations.getSingleLineDefaultConfiguration)(intlProps, {
      required: isRequired,
      size: editorSize
    });
  }
  if (hideLineNumbers) {
    return (0, import_getDefaultConfigurations.getDefaultConfiguration)(intlProps, {
      required: isRequired,
      size: editorSize
    });
  }
  return (0, import_getDefaultConfigurations.getMultipleLinesDefaultConfiguration)(intlProps, {
    gutterConfiguration,
    required: isRequired,
    size: editorSize
  });
};
function useLazyEditorViewRef() {
  const viewRef = (0, import_react.useRef)(null);
  if (viewRef.current === null) {
    viewRef.current = new import_view.EditorView({
      state: import_state.EditorState.create()
    });
  }
  return viewRef;
}
const getDomEventHandlers = (onFocus, onBlur) => {
  return import_view.EditorView.domEventHandlers({
    // Prevent the Escape key from propagating beyond the editor.
    // This is necessary for editors that are placed inside an overlay (e.g. Modal),
    // where pressing `Escape` to exit the edit mode would otherwise close the overlay.
    keydown: (event) => {
      if (event.key === "Escape") {
        event.stopPropagation();
      }
    },
    // when the editor content element receives focus, call the provided focus handler
    focus: (event) => {
      onFocus?.(event);
    },
    // when the editor content element loses focus, call the provided blur handler
    blur: (event) => {
      onBlur?.(event);
    }
  });
};
const BaseCodeEditor = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    controlState,
    value,
    readOnly = false,
    spellCheck = false,
    lineWrap = false,
    onChange,
    onFoldingChange,
    defaultFolding,
    folding,
    fullHeight = false,
    placeholder,
    configuration,
    singleLine = false,
    hideLineNumbers = false,
    size = "default",
    onFocus,
    onBlur,
    gutterConfiguration,
    id: propId,
    required = false,
    style: consumerStyle,
    className: consumerClassName,
    "aria-label": ariaLabel,
    "aria-labelledby": ariaLabelledBy,
    "aria-details": ariaDetails,
    "aria-describedby": ariaDescribedBy,
    "data-testid": dataTestId = import_testIds.baseCodeEditorRootTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    // @ts-expect-error - We want to allow overriding data-dt-component as we use the component internally in other Strato components, but it should not appear in the public API.
    "data-dt-component": dataDtComponent,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const intl = (0, import_react_intl.useIntl)();
  const [hasError, hint] = (0, import_useFormControlState.useFormControlState)(controlState);
  const containerRef = (0, import_react.useRef)(null);
  const rootRef = (0, import_useMergeRefs.useMergeRefs)([containerRef]);
  const sanitizedOnChange = (0, import_react.useCallback)(
    (value2) => onChange?.((0, import_sanitize_text.sanitizeText)(value2)),
    [onChange]
  );
  const viewRef = useLazyEditorViewRef();
  const { id } = (0, import_useSyncedFormControlProps.useSyncedFormControlProps)({
    propId
  });
  const valueChangeRef = (0, import_react.useRef)(void 0);
  const valueRef = (0, import_react.useRef)(value ?? "");
  const isMarkedDirtyRef = (0, import_react.useRef)(false);
  const prevEditorConfigRef = (0, import_react.useRef)(void 0);
  const prevLangConfigRef = (0, import_react.useRef)(void 0);
  const [previousValue, setPreviousValue] = (0, import_react.useState)("");
  const transactionRef = (0, import_react.useRef)(void 0);
  const { focusProps, isFocusVisible } = (0, import_useFocusRing.useFocusRing)({
    color: hasError ? "critical" : "neutral"
  });
  const [controlledFolding] = (0, import_useControllableState.useControllableState)({
    defaultValue: defaultFolding,
    onChange: (values) => {
      if (values) {
        onFoldingChange?.(values);
      }
    },
    value: folding
  });
  (0, import_react.useImperativeHandle)(
    forwardedRef,
    () => ({
      element: containerRef.current,
      editorView: () => viewRef.current,
      focus: () => viewRef.current.focus(),
      selectRange: (range) => {
        const editorView = viewRef.current;
        if (editorView && range) {
          const docLength = editorView.state.doc.length;
          const from = range.from < docLength ? range.from : 0;
          const to = range.to < docLength ? range.to : docLength;
          const transaction = editorView.state.update({
            selection: import_state.EditorSelection.range(to, from),
            scrollIntoView: true
          });
          editorView.dispatch(transaction);
          editorView.focus();
        }
      },
      setCursorPosition: (position) => {
        viewRef.current.dispatch({
          selection: import_state.EditorSelection.cursor(
            typeof position === "number" ? Math.min(
              Math.max(position, 0),
              viewRef.current.state.doc.length
            ) : viewRef.current.state.doc.length
          ),
          scrollIntoView: true
        });
        viewRef.current.focus();
      }
    }),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    []
  );
  (0, import_react.useEffect)(() => {
    const editorView = viewRef.current;
    containerRef.current?.appendChild(editorView.dom);
    editorView.dom.setAttribute(import_testIds.dataTestId, import_testIds.baseCodeEditorTestId);
    return () => editorView.destroy();
  }, []);
  (0, import_react.useEffect)(() => {
    const { contentDOM } = viewRef.current;
    const ariaLabelAttribute = "aria-label";
    const ariaLabelledByAttribute = "aria-labelledby";
    const ariaDetailsAttribute = "aria-details";
    const ariaDescribedByAttribute = "aria-describedby";
    if (ariaDescribedBy) {
      contentDOM.setAttribute(ariaDescribedByAttribute, ariaDescribedBy);
    }
    if (ariaDetails) {
      contentDOM.setAttribute(ariaDetailsAttribute, ariaDetails);
    }
    if (ariaLabelledBy) {
      contentDOM.setAttribute(ariaLabelledByAttribute, ariaLabelledBy);
    } else {
      contentDOM.setAttribute(
        ariaLabelAttribute,
        ariaLabel ?? intl.formatMessage({
          defaultMessage: "Code Editor",
          id: "EBcYgbL0S1T6EpS2",
          description: "Aria-label for the code editors."
        })
      );
    }
    return () => [
      ariaLabelAttribute,
      ariaLabelledByAttribute,
      ariaDetailsAttribute,
      ariaDescribedByAttribute
    ].forEach((ariaAttribute) => contentDOM.removeAttribute(ariaAttribute));
  }, [ariaLabel, ariaLabelledBy, ariaDetails, ariaDescribedBy, intl]);
  (0, import_react.useLayoutEffect)(() => {
    viewRef.current.requestMeasure();
  }, [viewRef, size]);
  const { messages, defaultLocale, locale } = (0, import_react_intl.useIntl)();
  (0, import_react.useEffect)(() => {
    const editorView = viewRef.current;
    const handleValueChange = (transaction) => {
      const newValue = transaction.newDoc.toString();
      if (valueRef.current !== newValue) {
        transactionRef.current = transaction;
        valueRef.current = newValue;
        if (isMarkedDirtyRef.current) {
          isMarkedDirtyRef.current = false;
        } else {
          valueChangeRef.current?.(newValue);
        }
      } else {
        isMarkedDirtyRef.current = false;
      }
      return transaction;
    };
    editorView.setState(
      import_state.EditorState.create({
        // Initial value of the editor
        doc: value ?? "",
        extensions: [
          getDomEventHandlers(onFocus, onBlur),
          import_state.Prec.highest(keymapCompartment.of((0, import_key_maps.getKeyMaps)(viewRef))),
          languageCompartment.of([]),
          editorBaseConfigCompartment.of([]),
          // Indent unit is two spaces instead of tab
          import_language.indentUnit.of("  "),
          // Filter for transactions that gets applied before they are processed. This allows us
          // to control which transactions go through as well as adding/editing transactions if we please.
          import_state.EditorState.transactionFilter.of((transaction) => {
            if (transaction.docChanged) {
              return handleValueChange(transaction);
            }
            return transaction;
          }),
          import_view.EditorView.updateListener.of((update) => {
            const isUndoOrRedo = update.transactions.some(
              (transaction) => transaction.isUserEvent("undo") || transaction.isUserEvent("redo")
            );
            if (update.geometryChanged && !update.docChanged && !update.focusChanged && update.heightChanged) {
              const ranges = (0, import_language.foldedRanges)(update.state);
              const foldValues = (0, import_calculate_folding.calculateFolding)(ranges);
              onFoldingChange?.(foldValues);
            } else if (update.docChanged && isUndoOrRedo) {
              update.transactions.forEach(
                (transaction) => handleValueChange(transaction)
              );
            }
          }),
          lineWrappingCompartment.of([]),
          placeholderCompartment.of([]),
          highlightOnChangeCompartment.of([])
        ]
      })
    );
  }, []);
  (0, import_react.useEffect)(() => {
    const editorView = viewRef.current;
    const prev = prevEditorConfigRef.current;
    if (prev !== void 0 && prev.singleLine === singleLine && prev.hideLineNumbers === hideLineNumbers && prev.gutterConfiguration === gutterConfiguration && prev.required === required && prev.size === size && prev.locale === locale && prev.defaultLocale === defaultLocale && prev.messages === messages) {
      return;
    }
    prevEditorConfigRef.current = {
      singleLine,
      hideLineNumbers,
      gutterConfiguration,
      required,
      size,
      locale,
      defaultLocale,
      messages
    };
    const intlProps = { locale, defaultLocale, messages };
    editorView.dispatch({
      effects: editorBaseConfigCompartment.reconfigure(
        getEditorConfiguration(intlProps, {
          singleLine,
          hideLineNumbers,
          gutterConfiguration,
          required,
          size
        })
      )
    });
  }, [
    singleLine,
    hideLineNumbers,
    gutterConfiguration,
    required,
    size,
    locale,
    defaultLocale,
    messages
  ]);
  (0, import_react.useEffect)(() => {
    const editorView = viewRef.current;
    const prev = prevLangConfigRef.current;
    if (prev !== void 0 && prev.language === configuration?.language && prev.extendWith === configuration?.extendWith && prev.autocompleteOptions === configuration?.autocompleteOptions && prev.locale === locale && prev.defaultLocale === defaultLocale && prev.messages === messages) {
      return;
    }
    prevLangConfigRef.current = {
      language: configuration?.language,
      extendWith: configuration?.extendWith,
      autocompleteOptions: configuration?.autocompleteOptions,
      locale,
      defaultLocale,
      messages
    };
    const intlProps = { locale, defaultLocale, messages };
    editorView.dispatch({
      effects: languageCompartment.reconfigure(
        (0, import_language_extensions.getLanguageConfiguration)(configuration || {}, intlProps)
      )
    });
  }, [configuration, locale, defaultLocale, messages]);
  (0, import_react.useEffect)(() => {
    (0, import_configure_read_only.configureReadOnly)(readOnly, viewRef, keymapCompartment);
  }, [readOnly]);
  (0, import_react.useEffect)(() => {
    (0, import_configure_line_wrap.configureLineWrap)(lineWrap, viewRef, lineWrappingCompartment);
  }, [lineWrap]);
  (0, import_react.useEffect)(() => {
    (0, import_configure_placeholder.configurePlaceholder)(placeholder, viewRef, placeholderCompartment);
  }, [placeholder]);
  (0, import_react.useEffect)(() => {
    (0, import_configure_spell_check.configureSpellCheck)(spellCheck, readOnly, viewRef);
  }, [viewRef, readOnly, spellCheck]);
  (0, import_react.useEffect)(() => {
    valueChangeRef.current = onChange !== void 0 ? sanitizedOnChange : void 0;
  }, [onChange, sanitizedOnChange]);
  (0, import_react.useEffect)(() => {
    const sanitizedValue = (0, import_sanitize_value.sanitizeValue)(value);
    if (viewRef.current.state.doc.toString() === sanitizedValue) {
      return;
    }
    setPreviousValue(viewRef.current.state.doc.toString());
    isMarkedDirtyRef.current = true;
    try {
      let selection;
      if (!viewRef.current.hasFocus) {
        selection = void 0;
      } else {
        const editorLineCount = viewRef.current.state.doc.lines;
        const lastLine = viewRef.current.state.doc.line(editorLineCount);
        const selectionPosition = Math.max(0, lastLine.length - 1);
        selection = { anchor: selectionPosition, head: selectionPosition };
      }
      transactionRef.current = void 0;
      const transaction = viewRef.current.state.update({
        changes: [
          {
            from: 0,
            to: viewRef.current.state.doc.length,
            insert: sanitizedValue
          }
        ],
        selection
      });
      viewRef.current.dispatch(transaction);
    } catch {
    }
  }, [value]);
  (0, import_react.useEffect)(() => {
    const editorView = viewRef.current;
    editorView.dispatch({
      effects: highlightOnChangeCompartment.reconfigure([
        (0, import_highlight_extension.highlightExtension)(previousValue)
      ])
    });
  }, [previousValue]);
  (0, import_react.useEffect)(() => {
    (0, import_run_foldings.runFoldings)(controlledFolding ?? [], viewRef.current);
  }, [controlledFolding]);
  (0, import_react.useEffect)(() => {
    const codeEditorInput = viewRef.current.dom.querySelector(
      "div[contenteditable]"
    );
    if (codeEditorInput) {
      hasError ? codeEditorInput.setAttribute("aria-invalid", "true") : codeEditorInput.removeAttribute("aria-invalid");
      codeEditorInput.tabIndex = -1;
    }
  }, [hasError]);
  const handleOutsideEditorClick = (0, import_react.useCallback)(
    (e) => {
      const autoCompleteContainer = document.body.querySelector(
        ".cm-tooltip-autocomplete"
      );
      if (!autoCompleteContainer?.contains(e.target)) {
        (0, import_autocomplete.closeCompletion)(viewRef.current);
      }
    },
    [viewRef]
  );
  (0, import_react.useEffect)(() => {
    document.addEventListener("pointerdown", handleOutsideEditorClick);
    return () => {
      document.removeEventListener("pointerdown", handleOutsideEditorClick);
    };
  }, [handleOutsideEditorClick]);
  (0, import_react.useEffect)(() => {
    if (!value) {
      return;
    }
    let timeoutId = null;
    let done = false;
    const cleanup = () => {
      if (timeoutId) {
        window.clearTimeout(timeoutId);
      }
    };
    const tick = () => {
      const wholeSyntaxTreeParsed = (0, import_language.syntaxTreeAvailable)(
        viewRef.current.state,
        viewRef.current.state.doc.length
      );
      if (wholeSyntaxTreeParsed) {
        cleanup();
        return;
      }
      done = (0, import_language.forceParsing)(
        viewRef.current,
        viewRef.current.state.doc.length,
        100
      );
      if (!done) {
        timeoutId = window.setTimeout(tick, 10);
      }
    };
    tick();
    return cleanup;
  }, [value]);
  (0, import_react.useEffect)(() => {
    const editorScrollContainer = viewRef.current.dom.querySelector(".cm-scroller");
    editorScrollContainer?.setAttribute("tabIndex", "0");
    editorScrollContainer?.setAttribute("role", "group");
    editorScrollContainer?.setAttribute(
      import_testIds.dataTestId,
      import_testIds.baseCodeEditorScrollerTestId
    );
    const handleFocus = (event) => {
      if (event instanceof FocusEvent) {
        focusProps.onFocus(event);
      }
    };
    const handleBlur = (event) => {
      if (event instanceof FocusEvent) {
        focusProps.onBlur(event);
      }
    };
    const handleKeyDown = (event) => {
      if (event instanceof KeyboardEvent) {
        if (event.key === "Enter" && !event.metaKey && !event.ctrlKey && !readOnly) {
          event.preventDefault();
          viewRef.current.focus();
        }
      }
    };
    editorScrollContainer?.addEventListener("focus", handleFocus);
    editorScrollContainer?.addEventListener("blur", handleBlur);
    editorScrollContainer?.addEventListener("keydown", handleKeyDown);
    return () => {
      editorScrollContainer?.removeEventListener("focus", handleFocus);
      editorScrollContainer?.removeEventListener("blur", handleBlur);
      editorScrollContainer?.removeEventListener("keydown", handleKeyDown);
    };
  }, [focusProps, readOnly, viewRef]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "div",
      {
        ref: rootRef,
        className: (0, import_clsx.default)(
          styles.editorContainer({
            focusVisible: isFocusVisible,
            fullHeight,
            hasError,
            lineNumbers: singleLine || hideLineNumbers ? "hidden" : void 0,
            readOnly,
            size
          }),
          import_InputGroup_css.inputGroupClassName,
          consumerClassName
        ),
        style: consumerStyle,
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        ...behavioralTrackingProps,
        ...restProps,
        "data-dt-component": dataDtComponent
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "input",
      {
        type: "text",
        id,
        disabled: readOnly,
        tabIndex: -1,
        onFocus: () => (
          // forward focus to the editor
          // we need the hidden input unfortunately since the editor does not support focus properly
          viewRef.current.focus()
        ),
        "aria-hidden": "true",
        style: {
          position: "absolute",
          height: 0,
          width: 0,
          opacity: 0,
          margin: 0,
          padding: 0,
          border: 0
        }
      }
    ),
    (hasError || hint) && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "span",
      {
        "aria-live": "polite",
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_FormFieldMessageBaseItem.InternalBaseItem, { variant: hasError ? "error" : "hint", children: hint })
      }
    )
  ] });
});
BaseCodeEditor.displayName = "BaseCodeEditor";
