// @dynatrace/strato-components/esm/charts/tree-map/components/TreeMapToolbar/TreeMapDownloadDataButton.js
import { jsx } from "react/jsx-runtime";
import { useRef } from "react";
import { useIntl } from "react-intl";
import { DownloadIcon } from "@dynatrace/strato-icons";
import { ChartToolbarRenderer } from "../../../core/components/toolbar/ChartToolbarRenderer.js";
import { TOOLBAR_MESSAGES } from "../../../core/components/toolbar/messages.js";
import { useA11yToolbarTabIndex } from "../../../core/hooks/useA11yToolbarTabIndex.js";
import { useTreeMapRawData } from "../../context/TreeMapRawDataContext.js";
import { buildAndDownloadCsv } from "../../utils/build-and-download-csv.js";
const TreeMapDownloadDataButton = () => {
  const downloadDataButtonRef = useRef(null);
  const intl = useIntl();
  const tabIndex = useA11yToolbarTabIndex("download");
  const { tree } = useTreeMapRawData();
  const handleDownloadDataKeyboard = (event) => {
    if (event.key === "Enter") {
      buildAndDownloadCsv(tree);
    }
  };
  const handleDownloadDataToolbar = () => {
    buildAndDownloadCsv(tree);
  };
  return /* @__PURE__ */ jsx(
    ChartToolbarRenderer.Control,
    {
      ref: downloadDataButtonRef,
      prefixIcon: /* @__PURE__ */ jsx(DownloadIcon, {}),
      tooltipText: intl.formatMessage(TOOLBAR_MESSAGES.downloadData),
      testId: "map-toolbar-download",
      onKeyDown: handleDownloadDataKeyboard,
      onClick: handleDownloadDataToolbar,
      tabIndex
    }
  );
};
export {
  TreeMapDownloadDataButton
};
