// @dynatrace/strato-components/esm/forms/forms-core/components/label/Label.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import { forwardRef } from "react";
import { FormattedMessage } from "react-intl";
import * as styles from "./Label.sty.js";
import { useBehavioralTrackingProps } from "../../../../core/hooks/useBehavioralTrackingProps.js";
import { canReceiveDefaultTextFormat as _canReceiveDefaultTextFormat } from "../../../../core/utils/can-receive-default-text-format.js";
import { mergeProps } from "../../../../core/utils/merge-props.js";
import { Text } from "../../../../typography/text/Text.js";
import { useFormFieldContext } from "../../hooks/useFormFieldContext.js";
const COMPONENT_NAME = "Label";
const SimpleLabelText = forwardRef(
  (props, forwardedRef) => /* @__PURE__ */ jsx(
    Text,
    {
      ref: forwardedRef,
      as: "label",
      textStyle: "small-emphasized",
      ...props
    }
  )
);
const Label = /* @__PURE__ */ forwardRef(
  (props, forwardedRef) => {
    const {
      id,
      children,
      htmlFor,
      disabled: propDisabled,
      required: propRequired,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "aria-disabled": ariaDisabledProp,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
    const {
      controlId: formFieldContextId,
      required,
      disabled,
      ariaDisabled,
      readOnly
    } = useFormFieldContext();
    const isStringChild = _canReceiveDefaultTextFormat(children);
    const Component = isStringChild ? SimpleLabelText : "label";
    const labelFor = htmlFor || formFieldContextId;
    return /* @__PURE__ */ jsxs(
      Component,
      {
        htmlFor: labelFor,
        id,
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        ref: forwardedRef,
        className: clsx(
          styles.label({
            disabled: disabled || propDisabled || ariaDisabled,
            // Only apply the required styles if we have string children, otherwise block level
            // elements in children will move the `*` required marker somewhere.
            required: isStringChild && (required ?? propRequired),
            readOnly
          }),
          consumerClassName
        ),
        style: consumerStyle,
        ...mergeProps(restProps, behavioralTrackingProps),
        "data-dt-component": COMPONENT_NAME,
        children: [
          children,
          isStringChild && readOnly && /* @__PURE__ */ jsx(Text, { textStyle: "small", className: clsx(styles.readOnlyLabel), children: /* @__PURE__ */ jsx(
            FormattedMessage,
            {
              defaultMessage: "(read only)",
              id: "hIBEZw/eT6Nb+dz3",
              description: "Text appended to a label when the associated form field is read-only."
            }
          ) })
        ]
      }
    );
  }
);
Label.displayName = COMPONENT_NAME;
export {
  Label
};
