// @dynatrace/strato-components/esm/content/microguide/Microguide.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import {
  forwardRef,
  useEffect,
  useLayoutEffect,
  useRef,
  useState
} from "react";
import { createPortal } from "react-dom";
import { FormattedMessage, useIntl } from "react-intl";
import Colors from "@dynatrace/strato-design-tokens/colors";
import { MicroguideContext } from "./contexts/MicroguideContext.js";
import * as styles from "./Microguide.sty.js";
import { Step } from "./Step.js";
import { Button } from "../../buttons/button/Button.js";
import { useAriaLabelingProps } from "../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useOverlayStacking } from "../../core/hooks/useOverlayStacking.js";
import { useSafeForwardProps } from "../../core/hooks/useSafeForwardProps.js";
import { useSlots } from "../../core/hooks/useSlots.js";
import { Flex } from "../../layouts/flex/Flex.js";
import { Surface } from "../../layouts/surface/Surface.js";
import { useKeyboardFocusHandling } from "../../overlays/hooks/useKeyboardFocusHandling.js";
import { Content } from "../shared-components/Content.js";
import { Title } from "../shared-components/Title.js";
import { Visual } from "../shared-components/Visual.js";
const COMPONENT_NAME = "Microguide";
const MicroguideRoot = forwardRef(
  (props, forwardedRef) => {
    const {
      open,
      placement = "bottom-right",
      onClose,
      children,
      onStepChange,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      className: consumerClassName,
      style: consumerStyle,
      ...remainingProps
    } = props;
    const microguideContainer = useRef(null);
    const closeButtonRef = useRef(null);
    const [currentStep, setCurrentStep] = useState(0);
    const [isNext, setIsNext] = useState(true);
    const [hasTitle, setHasTitle] = useState(false);
    const [hasContent, setHasContent] = useState(false);
    const [handleFocusOnKeydown, closedOnKeydown] = useKeyboardFocusHandling(closeButtonRef);
    const overlayStacking = useOverlayStacking(void 0, "microguide");
    const intl = useIntl();
    const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
    const ariaLabelingProps = useAriaLabelingProps(restProps);
    const forwardProps = useSafeForwardProps(restProps, ariaLabelingProps);
    const [slots] = useSlots(children, {
      steps: { match: Step, multiple: true }
    });
    const stepsLength = slots.steps?.length ?? 0;
    const isLastStep = currentStep + 1 >= stepsLength;
    const nextGotItDataDtProperties = `state:${isLastStep ? "gotIt" : "next"}`;
    if (hasTitle && ariaLabelingProps["aria-labelledby"] === void 0) {
      ariaLabelingProps["aria-labelledby"] = "strato-microguide-title";
    }
    if (hasContent && ariaLabelingProps["aria-describedby"] === void 0) {
      ariaLabelingProps["aria-describedby"] = "strato-microguide-content";
    }
    const closeOnEscape = (event) => {
      if (event.key !== "Escape") {
        return;
      }
      event.preventDefault();
      closedOnKeydown.current = true;
      onClose(currentStep, isLastStep);
    };
    useLayoutEffect(() => {
      const container = document.createElement("div");
      container.setAttribute("aria-live", "polite");
      container.setAttribute("data-microguide-container", "true");
      microguideContainer.current = container;
      const toastContainer = document.querySelector("[data-toast-container]");
      if (toastContainer?.parentNode) {
        toastContainer.parentNode.insertBefore(container, toastContainer);
      } else {
        document.body.appendChild(container);
      }
      return () => {
        microguideContainer.current?.remove();
      };
    }, []);
    useEffect(() => {
      if (open === false) {
        setCurrentStep(0);
        handleFocusOnKeydown();
      }
    }, [handleFocusOnKeydown, open]);
    const handleBack = () => {
      const previousStep = currentStep - 1;
      setIsNext(false);
      onStepChange?.(previousStep);
      setCurrentStep(previousStep);
    };
    const handleNext = () => {
      const nextStep = currentStep + 1;
      setIsNext(true);
      onStepChange?.(nextStep);
      setCurrentStep(nextStep);
    };
    return open ? createPortal(
      /* @__PURE__ */ jsxs(
        Surface,
        {
          ref: forwardedRef,
          elevation: "floating",
          onKeyDown: closeOnEscape,
          role: "dialog",
          "aria-modal": "false",
          className: clsx(
            styles.wrapper,
            styles.wrapperPosition[placement],
            consumerClassName
          ),
          style: {
            // since the surface applies `position: relative` by default,
            // we ensure higher specificity by applying position as an inline style
            position: "fixed",
            zIndex: overlayStacking.style.zIndex,
            ...consumerStyle
          },
          "data-testid": dataTestId,
          "data-dtrum-mask": dataDtrumMask,
          "data-dtrum-allow": dataDtrumAllow,
          ...ariaLabelingProps,
          ...behavioralTrackingProps,
          ...forwardProps,
          "data-dt-component": COMPONENT_NAME,
          children: [
            /* @__PURE__ */ jsx(
              MicroguideContext.Provider,
              {
                value: {
                  onClose,
                  currentStep,
                  totalSteps: stepsLength,
                  isNext,
                  setHasContent,
                  setHasTitle,
                  closeButtonRef,
                  closedOnKeydown
                },
                children: slots.steps?.[currentStep]
              }
            ),
            /* @__PURE__ */ jsxs(
              Flex,
              {
                className: styles.footer,
                flexDirection: "row",
                justifyContent: "space-between",
                alignItems: "center",
                gap: 16,
                paddingX: 16,
                paddingBottom: 16,
                paddingTop: 8,
                children: [
                  /* @__PURE__ */ jsx("div", { style: { color: Colors.Text.Neutral.Subdued }, children: stepsLength > 1 && `${currentStep + 1}/${stepsLength}` }),
                  /* @__PURE__ */ jsxs(Flex, { gap: 8, children: [
                    currentStep > 0 && /* @__PURE__ */ jsx(Button, { onClick: handleBack, "data-dt-name": "Back", children: /* @__PURE__ */ jsx(
                      FormattedMessage,
                      {
                        description: "Microguide Back button.",
                        defaultMessage: "Back",
                        id: "N27hQ6LyGqdym2Yi"
                      }
                    ) }),
                    /* @__PURE__ */ jsx(
                      Button,
                      {
                        variant: isLastStep ? "accent" : "emphasized",
                        color: isLastStep ? "primary" : "neutral",
                        onClick: isLastStep ? () => onClose(currentStep, true) : handleNext,
                        onKeyDown: () => closedOnKeydown.current = isLastStep,
                        "data-dt-name": "NextGotIt",
                        "data-dt-properties": nextGotItDataDtProperties,
                        children: isLastStep ? intl.formatMessage({
                          defaultMessage: "Got it",
                          id: "fzkORq3AsvX/NtMB",
                          description: "Microguide Got it button at last step."
                        }) : intl.formatMessage({
                          defaultMessage: "Next",
                          id: "+jn6HNnzB4IDIbZu",
                          description: "Microguide Next button."
                        })
                      }
                    )
                  ] })
                ]
              }
            )
          ]
        }
      ),
      microguideContainer.current ?? document.body
    ) : null;
  }
);
MicroguideRoot.displayName = COMPONENT_NAME;
const Microguide = Object.assign(MicroguideRoot, {
  Step,
  Title,
  Visual,
  Content
});
export {
  Microguide
};
