// @dynatrace/strato-components/esm/content/microguide/Step.js
import { jsx, jsxs } from "react/jsx-runtime";
import { createElement } from "react";
import {
  AnimatePresence,
  motion,
  useReducedMotion
} from "framer-motion";
import { useEffect } from "react";
import { useIntl } from "react-intl";
import AnimationsJS from "@dynatrace/strato-design-tokens/animations-js";
import { XmarkIcon } from "@dynatrace/strato-icons";
import { useMicroguideContext } from "./hooks/useMicroguideContext.js";
import * as styles from "./Microguide.sty.js";
import { Button } from "../../buttons/button/Button.js";
import { useSlots } from "../../core/hooks/useSlots.js";
import { moveEasing } from "../../core/utils/framer-motion-easings.js";
import { Flex } from "../../layouts/flex/Flex.js";
import { Content } from "../shared-components/Content.js";
import { Title } from "../shared-components/Title.js";
import { Visual } from "../shared-components/Visual.js";
const contentVariants = {
  enter: (isNext) => {
    return {
      x: isNext ? "100%" : "-100%",
      opacity: 0
    };
  },
  center: {
    x: 0,
    opacity: 1,
    position: "relative"
  },
  exit: (isNext) => {
    return {
      x: isNext ? "-100%" : "100%",
      opacity: 0,
      position: "absolute",
      left: 0
    };
  }
};
const fadeProps = {
  initial: {
    opacity: 0
  },
  animate: {
    opacity: 1,
    position: "relative"
  },
  exit: {
    opacity: 0,
    position: "absolute",
    left: 0
  }
};
const Step = ({ children }) => {
  const {
    onClose,
    currentStep,
    totalSteps,
    isNext,
    setHasContent,
    setHasTitle,
    closeButtonRef,
    closedOnKeydown
  } = useMicroguideContext();
  const intl = useIntl();
  const shouldReduceMotion = useReducedMotion();
  const transitionProp = {
    duration: shouldReduceMotion ? 0 : AnimationsJS.SlideIn.Duration.Base,
    ease: moveEasing
  };
  const [slots] = useSlots(children, {
    title: Title,
    visual: Visual,
    content: Content
  });
  useEffect(() => {
    setHasTitle(slots.title !== void 0);
    setHasContent(slots.content !== void 0);
  }, [setHasContent, setHasTitle, slots.content, slots.title]);
  return /* @__PURE__ */ jsxs(
    Flex,
    {
      as: motion.div,
      flexDirection: "column",
      gap: 8,
      className: styles.stepContainer,
      children: [
        /* @__PURE__ */ jsxs(
          Flex,
          {
            flexDirection: "row",
            justifyContent: "space-between",
            alignItems: "flex-start",
            gap: 16,
            paddingTop: 8,
            paddingLeft: 16,
            paddingRight: 8,
            children: [
              /* @__PURE__ */ jsx(
                Flex,
                {
                  flexItem: true,
                  paddingY: 6,
                  flexGrow: 1,
                  id: "strato-microguide-title",
                  style: { position: "relative" },
                  children: /* @__PURE__ */ jsx(AnimatePresence, { initial: false, children: slots.title && /* @__PURE__ */ createElement(
                    motion.div,
                    {
                      ...fadeProps,
                      transition: transitionProp,
                      key: `strato-microguide-title-${currentStep}`
                    },
                    slots.title
                  ) })
                }
              ),
              /* @__PURE__ */ jsx(
                Button,
                {
                  onClick: () => onClose(currentStep, currentStep + 1 === totalSteps),
                  onKeyDown: () => closedOnKeydown.current = true,
                  ref: closeButtonRef,
                  "aria-label": intl.formatMessage({
                    defaultMessage: "Close microguide",
                    id: "0yl3EugW5oCHVbO1",
                    description: "Close the microguide."
                  }),
                  "data-dt-name": "CloseX",
                  children: /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(XmarkIcon, {}) })
                }
              )
            ]
          }
        ),
        slots.visual && /* @__PURE__ */ jsx("div", { className: styles.visualWrapper, children: /* @__PURE__ */ jsx(AnimatePresence, { initial: false, children: /* @__PURE__ */ createElement(
          motion.div,
          {
            ...fadeProps,
            transition: transitionProp,
            key: `strato-microguide-visual-${currentStep}`
          },
          slots.visual
        ) }) }),
        slots.content && /* @__PURE__ */ jsx("div", { className: styles.contentWrapper, id: "strato-microguide-content", children: /* @__PURE__ */ jsx(AnimatePresence, { initial: false, custom: isNext, children: /* @__PURE__ */ jsx(
          Flex,
          {
            as: motion.div,
            variants: contentVariants,
            className: styles.animatedContent,
            custom: isNext,
            initial: "enter",
            animate: "center",
            exit: "exit",
            transition: transitionProp,
            flexDirection: "column",
            children: slots.content
          },
          `strato-microguide-content-${currentStep}`
        ) }) })
      ]
    }
  );
};
Step["displayName"] = "Microguide.Step";
export {
  Step
};
