// @dynatrace/strato-components/esm/charts/annotations-chart/utils/annotations-chart-config.js
import { merge } from "lodash-es";
import { populateMissingTooltipProps } from "./populate-missing-tooltip-props.js";
import {
  DEFAULT_LABELS_WIDTH,
  DEFAULT_TEXT_OVERFLOW,
  DEFAULT_TOOLTIP_MAX_LINES,
  DEFAULT_TRUNCATION_MODE
} from "../../core/components/annotations/constants.js";
import { populateMissingAxisProps } from "../components/axis/utils/populate-missing-axis-props.js";
import { DEFAULT_ANNOTATIONS_CHART_CONFIG } from "../constants.js";
function deserializeConfig(serializedConfig) {
  try {
    return merge(
      {},
      DEFAULT_ANNOTATIONS_CHART_CONFIG,
      JSON.parse(serializedConfig)
    );
  } catch {
    return DEFAULT_ANNOTATIONS_CHART_CONFIG;
  }
}
function isSerializedConfig(contextConfig) {
  return typeof contextConfig === "string";
}
function buildAnnotationsChartConfig(props, contextConfig, xAxis, tooltip, hasNumericalData) {
  return {
    showLabels: props.showLabels ?? contextConfig.showLabels ?? DEFAULT_ANNOTATIONS_CHART_CONFIG.showLabels,
    labelWidth: props.labelWidth ?? contextConfig.labelWidth ?? DEFAULT_LABELS_WIDTH,
    height: props.height ?? contextConfig.height,
    textOverflow: props.textOverflow ?? contextConfig?.textOverflow ?? DEFAULT_TEXT_OVERFLOW,
    truncateMode: props.truncateMode ?? contextConfig?.truncateMode ?? DEFAULT_TRUNCATION_MODE,
    maxLines: props.maxLines ?? contextConfig?.maxLines ?? DEFAULT_TOOLTIP_MAX_LINES,
    xAxisConfig: populateMissingAxisProps(
      xAxis,
      hasNumericalData,
      contextConfig.xAxisConfig
    ),
    tooltipConfig: populateMissingTooltipProps(
      tooltip,
      contextConfig.tooltipConfig
    )
  };
}
export {
  buildAnnotationsChartConfig,
  deserializeConfig,
  isSerializedConfig
};
