// @dynatrace/strato-components/esm/charts/honeycomb/components/HoneycombChartLayout.js
import { jsx, jsxs } from "react/jsx-runtime";
import { isEmpty, isUndefined } from "lodash-es";
import { ChartLayout } from "../../core/components/chart-layout/ChartLayout.js";
import { useAutoLegendRefresh } from "../../core/components/chart-layout/hooks/useAutoLegendRefresh.js";
import { getChartStatesTemplates } from "../../core/components/chart-states/utils/get-chart-states-template.js";
import { DEFAULT_HONEYCOMB_WIDTH } from "../constants.js";
import { graphContainer } from "../HoneycombChart.sty.js";
import { HoneycombHive } from "./HoneycombHive.js";
import { useHoneycombData } from "../hooks/use-honeycomb-data.js";
import { useHoneycombLegendConfig } from "../hooks/use-honeycomb-legend-config.js";
import { HoneycombLegendRenderer } from "./legend/HoneycombLegendRenderer.js";
const HoneycombChartLayout = (props) => {
  const {
    data,
    width = DEFAULT_HONEYCOMB_WIDTH,
    loading = false,
    children,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow
  } = props;
  const filteredData = useHoneycombData();
  const legendConfig = useHoneycombLegendConfig();
  const legendRatio = legendConfig?.initialRatio ?? legendConfig?.ratio;
  const chartLayoutRef = useAutoLegendRefresh(data);
  const { emptyState, errorState } = getChartStatesTemplates(children);
  return /* @__PURE__ */ jsxs(
    ChartLayout,
    {
      ref: chartLayoutRef,
      chartWidth: width,
      chartHeight: "100%",
      isEmpty: isEmpty(filteredData),
      emptyState,
      errorState,
      showLoader: loading,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      children: [
        /* @__PURE__ */ jsx(ChartLayout.Graph, { children: /* @__PURE__ */ jsx("div", { className: graphContainer, children: /* @__PURE__ */ jsx(HoneycombHive, {}) }) }),
        !isUndefined(legendConfig) && !legendConfig.hidden && /* @__PURE__ */ jsx(
          ChartLayout.Legend,
          {
            position: legendConfig.position,
            initialRatio: legendRatio,
            onResize: legendConfig.onRatioChange,
            children: /* @__PURE__ */ jsx(HoneycombLegendRenderer, {})
          }
        )
      ]
    }
  );
};
HoneycombChartLayout["displayName"] = "HoneycombChartLayout";
export {
  HoneycombChartLayout
};
