// @dynatrace/strato-components/esm/filters/filter-field/hooks/recent-and-pinned/usePersistedFilter.js
import { useMutation, useQuery } from "@tanstack/react-query";
import { cloneDeep } from "lodash-es";
import { useContext, useState } from "react";
import { getCurrentUserDetails } from "@dynatrace-sdk/app-environment";
import {
  isForbidden,
  isNotFound,
  isUnauthorized,
  stateClient
} from "@dynatrace-sdk/client-state";
import { useRecentAndPinnedFiltersStore } from "./useRecentFiltersStore.js";
import { RecentAndSavedFiltersQueryClientContext } from "../../contexts/SavedAndRecentQueryClientContext.js";
import { useFilterFieldLoggerContext } from "../useFilterFieldLoggerContext.js";
const currentUser = getCurrentUserDetails();
function usePersistedFilter(name, key, saveFn) {
  const queryClient = useContext(RecentAndSavedFiltersQueryClientContext);
  const setCanPin = useRecentAndPinnedFiltersStore(
    (state) => state.setCanPin,
    name
  );
  const { logger } = useFilterFieldLoggerContext();
  return useQuery(
    {
      enabled: name !== null,
      queryKey: [key],
      throwOnError: false,
      queryFn: async ({ signal }) => {
        let savedFilters;
        try {
          savedFilters = await stateClient.getUserAppState({
            key,
            abortSignal: signal
          });
        } catch (error) {
          if (isUnauthorized(error) || isForbidden(error)) {
            logger?.warn(
              "FilterField pinned and recent error - permissions not met",
              error
            );
            setCanPin(false);
          } else if (signal.aborted) {
          } else if (!isNotFound(error)) {
            logger?.warn("FilterField pinned and recent error", error);
          }
          return [];
        }
        const parsedValue = JSON.parse(savedFilters.value, (key2, value) => {
          if (key2 === "timeUsed") {
            return new Date(value);
          }
          return value;
        });
        for (const parsedFilter of parsedValue) {
          if (!parsedFilter.timeUsed) {
            parsedFilter.timeUsed = /* @__PURE__ */ new Date();
          }
        }
        saveFn(parsedValue);
        return parsedValue;
      }
    },
    queryClient
  );
}
function usePersistedFilterMutation(key, name, actionType) {
  const queryClient = useContext(RecentAndSavedFiltersQueryClientContext);
  const { logger } = useFilterFieldLoggerContext();
  const setCanPin = useRecentAndPinnedFiltersStore(
    (state2) => state2.setCanPin,
    name
  );
  const state = useRecentAndPinnedFiltersStore((state2) => {
    if (actionType === "save") {
      return state2.recentFilters;
    }
    return state2.pinnedFilters;
  }, name);
  const setState = useRecentAndPinnedFiltersStore((state2) => {
    if (actionType === "save") {
      return state2.setRecentFilters;
    }
    return state2.setPinnedFilters;
  }, name);
  const action = useRecentAndPinnedFiltersStore((state2) => {
    if (actionType === "save") {
      return state2.saveRecentFilter;
    }
    if (actionType === "pin") {
      return state2.pinFilter;
    }
    return state2.unpinFilter;
  }, name);
  const [previousState, setPreviousState] = useState(
    cloneDeep(state)
  );
  return useMutation(
    {
      mutationKey: [key],
      throwOnError: false,
      mutationFn: async (filterString) => {
        if (name === null) {
          return Promise.resolve();
        }
        if (filterString.trim() === "") {
          return Promise.resolve();
        }
        setPreviousState(cloneDeep(state));
        const newState = action(filterString);
        await queryClient.cancelQueries({
          queryKey: [key]
        });
        try {
          await stateClient.setUserAppState({
            body: {
              validUntilTime: null,
              modificationInfo: {
                lastModifiedBy: currentUser.id,
                lastModifiedTime: /* @__PURE__ */ new Date()
              },
              value: JSON.stringify(newState)
            },
            key
          });
          return newState;
        } catch (error) {
          if (isUnauthorized(error) || isForbidden(error)) {
            logger?.warn(
              "FilterField pinned and recent error - permissions not met",
              error
            );
            setCanPin(false);
          } else {
            logger?.warn("FilterField pinned and recent error", error);
          }
          setState(previousState);
        }
        return state;
      },
      onError: (error) => {
        setState(previousState);
        queryClient.setQueryData([key], previousState);
        if (isUnauthorized(error) || isForbidden(error)) {
          logger?.warn(
            "FilterField pinned and recent error - permissions not met",
            error
          );
          setCanPin(false);
        }
      },
      onSuccess: (newState) => {
        if (newState !== void 0) {
          queryClient.setQueryData([key], newState);
        }
      }
    },
    queryClient
  );
}
export {
  usePersistedFilter,
  usePersistedFilterMutation
};
