// @dynatrace/strato-components/esm/filters/filter-field/hooks/recent-and-pinned/useRecentFilterSuggestions.js
import { jsx } from "react/jsx-runtime";
import { useEffect, useReducer, useRef } from "react";
import { HistoryIcon } from "@dynatrace/strato-icons";
import { useRecentFilters } from "./useRecentFilters.js";
import { TimeIndicator } from "../../../../core/components/time-indicator/TimeIndicator.js";
import { Text } from "../../../../typography/text/Text.js";
import { recentFiltersEffect } from "../../editor-state-configs/editor-config.js";
function useRecentFilterSuggestions(editorViewRef, filterNamespace = null) {
  const recentFilters = useRecentFilters(filterNamespace);
  const editorViewRefChange = useRef(editorViewRef.current);
  const [state, forceUpdate] = useReducer((x) => x + 1, 0);
  if (editorViewRefChange.current !== editorViewRef.current) {
    editorViewRefChange.current = editorViewRef.current;
    forceUpdate();
  }
  useEffect(() => {
    if (editorViewRef.current && recentFilters && recentFilters.length > 0) {
      const recentFiltersSuggestions = recentFilters.map(
        (storedFilter) => ({
          value: storedFilter.filter,
          insertionStrategy: "replace-all",
          icon: /* @__PURE__ */ jsx(HistoryIcon, {}),
          displayValue: storedFilter.filter,
          /* TODO
           * Discussed with @peter.affenzeller.
           * The details are by default rendered below the displayValue.
           * Only a container query moves it over to the right. Check with UX if this is
           */
          details: storedFilter.timeUsed && /* @__PURE__ */ jsx(Text, { textStyle: "small", children: /* @__PURE__ */ jsx(TimeIndicator, { time: storedFilter.timeUsed.getTime() }) })
        })
      );
      editorViewRef.current.dispatch({
        effects: [recentFiltersEffect.of(recentFiltersSuggestions)]
      });
    }
  }, [state, editorViewRef.current, JSON.stringify(recentFilters)]);
}
export {
  useRecentFilterSuggestions
};
