// @dynatrace/strato-components/charts/xy-chart/components/legend/XYLegend/XYLegend.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var XYLegend_exports = {};
__export(XYLegend_exports, {
  XYLegend: () => XYLegend
});
module.exports = __toCommonJS(XYLegend_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_getXYLegendActions = require("./utils/getXYLegendActions.js");
var import_XYLegendIcon = require("./XYLegendIcon.js");
var import_useChartLayoutLoading = require("../../../../core/components/chart-layout/hooks/useChartLayoutLoading.js");
var import_useChartLayoutOrientation = require("../../../../core/components/chart-layout/hooks/useChartLayoutOrientation.js");
var import_orientation = require("../../../../core/components/chart-layout/utils/orientation.js");
var import_ChartLegend = require("../../../../core/components/legend/categorical/ChartLegend.js");
var import_useTooltipReducer = require("../../../../core/components/overlay-tooltip/useTooltipReducer.js");
var import_build_categorical_items_from_scale = require("../../../../core/utils/build-categorical-items-from-scale.js");
var import_safe_get = require("../../../../core/utils/safe-get.js");
var import_useXYChartColorScales = require("../../../hooks/useXYChartColorScales.js");
var import_useXYChartTruncationMode = require("../../../hooks/useXYChartTruncationMode.js");
var import_reducer_filter_series = require("../../../store/reducer-filter-series.js");
var import_selectors = require("../../../store/selectors.js");
var import_xy_chart_internals = require("../../../types/xy-chart-internals.js");
var import_is_range_legend = require("../../../utils/is-range-legend.js");
var import_type_guards = require("../../../utils/type-guards.js");
const CHART_ICONS = {
  [import_xy_chart_internals.LINE_VARIANT]: import_XYLegendIcon.LINE_CHART_ICON_SHAPE,
  [import_xy_chart_internals.BAR_VARIANT]: import_XYLegendIcon.BAR_CHART_ICON_SHAPE,
  [import_xy_chart_internals.RECT_VARIANT]: import_XYLegendIcon.RECT_CHART_ICON_SHAPE,
  [import_xy_chart_internals.AREA_VARIANT]: import_XYLegendIcon.AREA_CHART_ICON_SHAPE
};
function getLegendItems(series, metadata, isMixedSeries, scaleColorMap, isCategoricalRectOnly) {
  return series.flatMap((series2) => {
    if (series2.type === import_xy_chart_internals.RECT_VARIANT) {
      if ((0, import_is_range_legend.isRangeLegend)(metadata, series2)) {
        return [];
      }
      const scale = (0, import_safe_get.safeGet)(scaleColorMap, series2.id);
      const overrideItems = (0, import_safe_get.safeGet)(metadata, series2).categoricalColorOverrides;
      const shape = isCategoricalRectOnly ? "circle" : CHART_ICONS[series2.type];
      const items = (0, import_build_categorical_items_from_scale.buildCategoricalItemsFromScale)(
        scale,
        overrideItems,
        //We need to compare exactly the same id when an action is selected
        //In RectSeries, categories are unique per series.
        (category) => `${series2.id}--${category}`
      );
      return items.map((item) => ({
        ...item,
        icon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYLegendIcon.XYLegendIcon, { color: item.color.Default, shape })
      }));
    } else {
      const { name, color, shape = "circle" } = (0, import_safe_get.safeGet)(metadata, series2);
      const iconShape = isMixedSeries && series2.type !== import_xy_chart_internals.DOT_VARIANT ? CHART_ICONS[series2.type] : shape;
      return {
        id: series2.id,
        color,
        name,
        icon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYLegendIcon.XYLegendIcon, { color: color.Default, shape: iconShape })
      };
    }
  });
}
function XYLegend({ series, seriesToMetadata }) {
  const dispatch = (0, import_useTooltipReducer.useTooltipReducer)();
  const { orientation: layoutOrientation } = (0, import_useChartLayoutOrientation.useChartLayoutOrientation)();
  const orientation = (0, import_orientation.getLegendOrientation)(layoutOrientation);
  const setState = (0, import_selectors.useSetState)();
  const isLoading = (0, import_useChartLayoutLoading.useChartLayoutLoading)();
  const scaleColorMap = (0, import_useXYChartColorScales.useXYChartColorScales)();
  const { truncationMode } = (0, import_useXYChartTruncationMode.useXYChartTruncationMode)();
  const firstSeries = series[0];
  const isCategoricalRectOnly = (0, import_type_guards.isRectSeries)(firstSeries) && series.every(import_type_guards.isRectSeries) && !(0, import_is_range_legend.isRangeLegend)(seriesToMetadata, firstSeries);
  const isMixedSeries = (0, import_react.useMemo)(() => {
    const typeSet = new Set(series.map((s) => s.type));
    return typeSet.size >= 2;
  }, [series]);
  const legendItems = (0, import_react.useMemo)(
    () => getLegendItems(
      series,
      seriesToMetadata,
      isMixedSeries,
      scaleColorMap,
      isCategoricalRectOnly
    ),
    // eslint-disable-next-line react-hooks/exhaustive-deps, react-hooks/preserve-manual-memoization
    [series, seriesToMetadata]
  );
  const legendActions = (0, import_getXYLegendActions.getXYLegendActions)(series, seriesToMetadata);
  const updateVisibleSeries = (0, import_react.useCallback)(
    (visibleItems) => {
      const visibleIds = new Set(visibleItems.map((item) => item.id));
      dispatch({ type: "RESET_TOOLTIP" })();
      setState((0, import_reducer_filter_series.reduceFilterSeries)(visibleIds));
    },
    [setState, dispatch]
  );
  const updateHighlightedSeriesId = (legendItemId) => {
    setState((prev) => ({ ...prev, highlightedSeriesId: legendItemId }));
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ChartLegend.ChartLegend,
    {
      items: legendItems,
      orientation,
      truncationMode,
      interactionsDisabled: isLoading,
      onSeriesAction: legendActions,
      onHover: (legendItem) => updateHighlightedSeriesId(legendItem?.id),
      onClick: updateVisibleSeries
    }
  );
}
