// @dynatrace/strato-components/esm/content/keyboard-shortcut/KeyboardShortcut.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import { forwardRef, useMemo } from "react";
import {
  keyboardShortcutKeyCSS,
  keyboardShortcutKeyWrapperCSS
} from "./KeyboardShortcut.sty.js";
import { useAriaLabelingProps } from "../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useSafeForwardProps } from "../../core/hooks/useSafeForwardProps.js";
import { useContainerColoring } from "../../layouts/container/Container.js";
const COMPONENT_NAME = "KeyboardShortcut";
const keyMappings = {
  meta: "Cmd",
  control: "Ctrl",
  escape: "Esc",
  altgraph: "Option",
  delete: "Del",
  arrowup: "\u2191",
  arrowdown: "\u2193",
  arrowright: "\u2192",
  arrowleft: "\u2190",
  plus: "+",
  minus: "-",
  multiply: "*",
  divide: "/"
};
function createKeyArrayFromString(keys) {
  return keys.split(" ")[0].split("+").filter((key) => key !== "").map(
    (key) => keyMappings[key.toLowerCase()] ?? `${key.charAt(0).toUpperCase()}${key.slice(1)}`
  );
}
const KeyboardShortcut = forwardRef(
  (props, forwardedRef) => {
    const {
      keys,
      variant = "default",
      disabled = false,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...remainingProps
    } = props;
    const keyArray = useMemo(() => createKeyArrayFromString(keys), [keys]);
    const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
    const ariaLabelingProps = useAriaLabelingProps(restProps);
    const forwardProps = useSafeForwardProps(restProps, ariaLabelingProps);
    const containerColoring = useContainerColoring();
    const onAccent = containerColoring.variant === "accent";
    return /* @__PURE__ */ jsx(
      "kbd",
      {
        ref: forwardedRef,
        "aria-disabled": disabled,
        className: clsx(keyboardShortcutKeyWrapperCSS, consumerClassName),
        style: consumerStyle,
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        ...ariaLabelingProps,
        ...behavioralTrackingProps,
        ...forwardProps,
        "data-dt-component": COMPONENT_NAME,
        children: keyArray.map((key, index) => /* @__PURE__ */ jsx(
          "kbd",
          {
            className: keyboardShortcutKeyCSS({ variant, onAccent, disabled }),
            children: key
          },
          `${key}-${index}`
        ))
      }
    );
  }
);
KeyboardShortcut.displayName = COMPONENT_NAME;
export {
  KeyboardShortcut,
  createKeyArrayFromString
};
