// @dynatrace/strato-components/esm/charts/core/components/annotations/components/canvas/utils/calculate-marker-width-and-content.js
import { checkMarkerFitCentered } from "./check-marker-fit-centered.js";
import { getDefaultValueMarkerPointerPosition } from "./get-default-value-marker-pointer-position.js";
import { ELLIPSIS_CHARACTER } from "../../../../../../../typography/utils.js";
import { charsFontWidth12pixelsSize500Map } from "../../../../text-ellipsis/chars-font-width-map.js";
import {
  calculateWidth,
  truncateString
} from "../../../../text-ellipsis/truncate-string.js";
import { MARKER_CONTENT_TO_BORDER_PADDING } from "../../../constants.js";
const MIN_TEXT_TO_RENDER = `W${ELLIPSIS_CHARACTER}W`;
const RANGE_MARKER_WITH_SYMBOL_MIN_WIDTH = 20;
function getAvailableWidth(width) {
  return width - MARKER_CONTENT_TO_BORDER_PADDING * 2;
}
function getSanitizedContent(content, availableWidth, truncateMode) {
  const minimumContentAllowedWidth = calculateWidth(
    MIN_TEXT_TO_RENDER,
    charsFontWidth12pixelsSize500Map
  );
  if (minimumContentAllowedWidth > availableWidth) {
    return "";
  }
  return truncateString(
    content,
    availableWidth,
    charsFontWidth12pixelsSize500Map,
    truncateMode
  );
}
function resolveConstrainedLayout(content, xPosition, markerWidth, trackLeftBoundary, trackRightBoundary, isRange, truncateMode) {
  const spaceToTheRight = trackRightBoundary - xPosition;
  const moreSpaceRight = spaceToTheRight > xPosition;
  const constrainedWidth = moreSpaceRight ? Math.min(spaceToTheRight, markerWidth) : xPosition - trackLeftBoundary;
  return {
    sanitizedContent: getSanitizedContent(
      content,
      getAvailableWidth(constrainedWidth),
      truncateMode
    ),
    calculatedMarkerWidth: constrainedWidth,
    valueMarkerOrientation: moreSpaceRight ? "left" : "right",
    calculatedX: !moreSpaceRight && isRange ? trackLeftBoundary : xPosition
  };
}
function resolveExpandedStringLayout(content, width, isRange, textOverflow, boundaries, truncateMode) {
  const {
    xPosition,
    trackLeftBoundary,
    trackRightBoundary,
    markerWidth,
    defaultPointer
  } = boundaries;
  if (isRange && textOverflow !== "expand") {
    return {
      sanitizedContent: getSanitizedContent(
        content,
        getAvailableWidth(width),
        truncateMode
      ),
      calculatedMarkerWidth: width,
      valueMarkerOrientation: defaultPointer,
      calculatedX: xPosition
    };
  }
  const canExpandCenter = checkMarkerFitCentered(
    xPosition,
    markerWidth,
    trackLeftBoundary,
    trackRightBoundary
  );
  if (canExpandCenter) {
    const adjustedX = isRange ? width / 2 + xPosition - markerWidth / 2 : xPosition;
    return {
      sanitizedContent: content,
      calculatedMarkerWidth: markerWidth,
      valueMarkerOrientation: defaultPointer,
      calculatedX: adjustedX
    };
  }
  if (xPosition - markerWidth >= trackLeftBoundary) {
    const adjustedX = !isRange && textOverflow === "expand" ? xPosition - markerWidth : xPosition;
    return {
      sanitizedContent: content,
      calculatedMarkerWidth: markerWidth,
      valueMarkerOrientation: "right",
      calculatedX: adjustedX
    };
  }
  if (xPosition + markerWidth <= trackRightBoundary) {
    return {
      sanitizedContent: content,
      calculatedMarkerWidth: markerWidth,
      valueMarkerOrientation: "left",
      calculatedX: xPosition
    };
  }
  return resolveConstrainedLayout(
    content,
    xPosition,
    markerWidth,
    trackLeftBoundary,
    trackRightBoundary,
    isRange,
    truncateMode
  );
}
function calculateMarkerWidthAndContent({
  trackWidth,
  trackLeftBoundary,
  content,
  width,
  x,
  isRange,
  textOverflow,
  truncateMode
}) {
  const trackRightBoundary = trackWidth + trackLeftBoundary;
  const xPosition = trackLeftBoundary + x;
  const defaultPointer = getDefaultValueMarkerPointerPosition(
    xPosition,
    trackLeftBoundary,
    trackRightBoundary
  );
  if (typeof content !== "string") {
    const sanitizedContent = isRange && RANGE_MARKER_WITH_SYMBOL_MIN_WIDTH > getAvailableWidth(width) ? "" : content;
    return {
      sanitizedContent,
      calculatedMarkerWidth: width,
      valueMarkerOrientation: defaultPointer,
      calculatedX: xPosition
    };
  }
  const textWidth = calculateWidth(content, charsFontWidth12pixelsSize500Map);
  const markerWidth = Math.max(
    textWidth + MARKER_CONTENT_TO_BORDER_PADDING * 2,
    width
  );
  if (markerWidth <= width) {
    return {
      sanitizedContent: content,
      calculatedMarkerWidth: width,
      valueMarkerOrientation: defaultPointer,
      calculatedX: xPosition
    };
  }
  const boundaries = {
    xPosition,
    trackLeftBoundary,
    trackRightBoundary,
    markerWidth,
    defaultPointer
  };
  return resolveExpandedStringLayout(
    content,
    width,
    isRange,
    textOverflow,
    boundaries,
    truncateMode
  );
}
export {
  calculateMarkerWidthAndContent
};
