// @dynatrace/strato-components/esm/content/avatar/Avatar.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import { forwardRef } from "react";
import {
  avatarAbbreviationCSS,
  avatarDisabledCSS,
  avatarWrapperCSS
} from "./Avatar.sty.js";
import { Label } from "./Label.js";
import { Subtitle } from "./Subtitle.js";
import { useAriaLabelingProps } from "../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useSafeForwardProps } from "../../core/hooks/useSafeForwardProps.js";
import { useSlots } from "../../core/hooks/useSlots.js";
import { focusRingStatic } from "../../core/styles/focusRingStatic.sty.js";
import { _sanitizeSymbol } from "../../core/utils/sanitize-symbol.js";
import { Flex } from "../../layouts/flex/Flex.js";
import { textStyle as textStyleCSS } from "../../styles/textStyle.sty.js";
const COMPONENT_NAME = "Avatar";
const InternalAvatar = /* @__PURE__ */ forwardRef(
  (props, ref) => {
    const {
      children,
      size = "default",
      abbreviation,
      disabled = false,
      as,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
    const ariaLabelingProps = useAriaLabelingProps(restProps);
    const forwardProps = useSafeForwardProps(restProps, ariaLabelingProps);
    const [slots] = useSlots(children, {
      label: Label,
      subtitle: Subtitle
    });
    const Abbreviation = as ?? "div";
    const remainingConsumerClassname = (consumerClassName ?? "").replace(
      /rovingFocusElement/g,
      ""
    );
    const hasRovingFocus = (consumerClassName ?? "").includes(
      "rovingFocusElement"
    );
    return /* @__PURE__ */ jsx(
      "div",
      {
        className: clsx(
          avatarWrapperCSS,
          textStyleCSS({ textStyle: "base" }),
          remainingConsumerClassname
        ),
        style: consumerStyle,
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        ...behavioralTrackingProps,
        "data-dt-component": COMPONENT_NAME,
        children: /* @__PURE__ */ jsxs(Flex, { gap: size === "small" ? 4 : 8, alignItems: "center", minWidth: 0, children: [
          /* @__PURE__ */ jsx(
            Abbreviation,
            {
              ref,
              className: clsx(
                {
                  [avatarDisabledCSS]: disabled,
                  rovingFocusElement: hasRovingFocus
                },
                avatarAbbreviationCSS({ size, disabled }),
                textStyleCSS({ textStyle: size === "small" ? "small" : "base" }),
                focusRingStatic({ disabled, when: "focus-within-visible" })
              ),
              ...ariaLabelingProps,
              ...forwardProps,
              children: _sanitizeSymbol(abbreviation, size === "default" ? 2 : void 0)
            }
          ),
          (slots.label || slots.subtitle) && /* @__PURE__ */ jsxs(Flex, { flexDirection: "column", gap: 0, minWidth: 0, children: [
            slots.label,
            slots.subtitle
          ] })
        ] })
      }
    );
  }
);
InternalAvatar.displayName = COMPONENT_NAME;
const Avatar = Object.assign(InternalAvatar, {
  Label,
  Subtitle
});
export {
  Avatar
};
