// @dynatrace/strato-components/esm/content/avatar/Label.js
import { jsx } from "react/jsx-runtime";
import { forwardRef } from "react";
import { canReceiveDefaultTextFormat as _canReceiveDefaultTextFormat } from "../../core/utils/can-receive-default-text-format.js";
import { Text } from "../../typography/text/Text.js";
const Label = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const {
    children,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId
  } = props;
  const childIsText = _canReceiveDefaultTextFormat(children);
  return childIsText ? /* @__PURE__ */ jsx(
    Text,
    {
      as: "div",
      textStyle: "base",
      ref: forwardedRef,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      children
    }
  ) : /* @__PURE__ */ jsx(
    "div",
    {
      "data-testid": dataTestId,
      ref: forwardedRef,
      className: consumerClassName,
      style: consumerStyle,
      children
    }
  );
});
Label.displayName = "Avatar.Label";
export {
  Label
};
