// @dynatrace/strato-components/esm/tables/DataTable/features/LineWrap/LineWrap.js
import { jsx } from "react/jsx-runtime";
import { functionalUpdate } from "@tanstack/react-table";
import { useIntl } from "react-intl";
import { WrapTextIcon, WrapTextOffIcon } from "@dynatrace/strato-icons";
import { Button } from "../../../../buttons/button/Button.js";
import { useControllableState } from "../../../../core/hooks/useControllableState.js";
import { Tooltip } from "../../../../overlays/tooltip/Tooltip.js";
import { useDataTableConfigElement } from "../../../hooks/useDataTableConfig.js";
import { useDataTable } from "../../contexts/DataTableContext.js";
import { dataTableMessages } from "../../messages.js";
function DataTableGlobalLineWrapAction() {
  return null;
}
const LineWrap = {
  createColumn(column, table) {
    column.toggleLineWrap = (value) => {
      if (!column.getCanLineWrap()) {
        return;
      }
      table.setColumnLineWrap((oldValue) => {
        const columnsStatesToUpdate = computeLineWrapStateUpdate(column, value);
        if (typeof oldValue === "boolean") {
          const columnIds = table.getAllLeafColumnsWithoutBuiltin().map((col) => col.id);
          const currentColumnsState = Object.fromEntries(columnIds.map((id) => [id, oldValue]));
          return {
            ...currentColumnsState,
            ...columnsStatesToUpdate
          };
        }
        return {
          ...oldValue,
          ...columnsStatesToUpdate
        };
      });
    };
    column.getIsLineWrapped = () => {
      const lineWrapState = table.getState().columnLineWrap;
      if (typeof lineWrapState === "boolean") {
        return lineWrapState;
      }
      const childColumns = column.columns;
      if (childColumns.length > 0) {
        return childColumns.every((child) => child.getIsLineWrapped());
      }
      return lineWrapState?.[column.id] ?? false;
    };
    column.getCanLineWrap = () => {
      return table.options.enableLineWrap ?? true;
    };
  },
  createTable(table) {
    table.setColumnLineWrap = (updater) => {
      const newValue = functionalUpdate(
        updater,
        table.getState().columnLineWrap ?? false
      );
      table.options.onColumnLineWrapChange?.(newValue);
    };
  }
};
function computeLineWrapStateUpdate(column, value) {
  const newValue = value ?? !column.getIsLineWrapped();
  const childColumns = column.columns;
  if (childColumns.length === 0) {
    return {
      [column.id]: newValue
    };
  }
  return Object.fromEntries(childColumns.map((child) => [child.id, newValue]));
}
function useColumnLineWrap(props, options) {
  const configLineWrap = useDataTableConfigElement("lineWrap");
  const [columnLineWrap, onColumnLineWrapChange] = useControllableState({
    value: props.lineWrap ?? configLineWrap,
    defaultValue: props.defaultLineWrap,
    onChange: props.onLineWrapChange
  });
  options.state = { ...options.state, columnLineWrap };
  options.onColumnLineWrapChange = onColumnLineWrapChange;
}
function DataTableGlobalLineWrapToolbarAction() {
  const table = useDataTable();
  const intl = useIntl();
  const lineWrapState = table.getState().columnLineWrap;
  const Icon = lineWrapState === true ? /* @__PURE__ */ jsx(WrapTextOffIcon, {}) : /* @__PURE__ */ jsx(WrapTextIcon, {});
  const lineWrapLabel = lineWrapState === true ? intl.formatMessage(dataTableMessages.disableLineWrapLabel) : intl.formatMessage(dataTableMessages.enableLineWrapLabel);
  const lineWrapDtProperty = `state:${lineWrapState === true ? "unwrap" : "wrap"}`;
  return /* @__PURE__ */ jsx(Tooltip, { text: lineWrapLabel, children: /* @__PURE__ */ jsx(
    Button,
    {
      "aria-label": lineWrapLabel,
      "data-dt-name": "ToolbarLineWrap",
      "data-dt-properties": lineWrapDtProperty,
      onClick: () => table.setColumnLineWrap((oldLineWrap) => {
        if (oldLineWrap === true) {
          return {};
        }
        return true;
      }),
      children: /* @__PURE__ */ jsx(Button.Prefix, { children: Icon })
    }
  ) });
}
export {
  DataTableGlobalLineWrapAction,
  DataTableGlobalLineWrapToolbarAction,
  LineWrap,
  useColumnLineWrap
};
