// @dynatrace/strato-components/layouts/split-layout/slots.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var slots_exports = {};
__export(slots_exports, {
  ContentSlot: () => ContentSlot,
  PanelSlot: () => PanelSlot,
  ResizeHandleSlot: () => ResizeHandleSlot,
  extractSlotsProps: () => extractSlotsProps
});
module.exports = __toCommonJS(slots_exports);
var import_react = require("react");
const PanelSlot = (0, import_react.forwardRef)(
  (_props, _ref) => null
);
const ContentSlot = (0, import_react.forwardRef)(
  (_props, _ref) => null
);
const ResizeHandleSlot = (0, import_react.forwardRef)(
  (_props, _ref) => null
);
function isPanelSlot(element) {
  return element.type === PanelSlot;
}
function isContentSlot(element) {
  return element.type === ContentSlot;
}
function isResizeHandleSlot(element) {
  return element.type === ResizeHandleSlot;
}
function extractSlotsProps(children) {
  let panel = null;
  let content = null;
  let resizeHandle = null;
  import_react.Children.forEach(children, (child) => {
    if (!(0, import_react.isValidElement)(child)) {
      return;
    }
    if (isPanelSlot(child)) {
      if (panel) {
        throw new Error("SplitLayout: only one Panel slot is allowed");
      }
      const {
        minWidth,
        maxWidth,
        defaultWidth,
        position,
        resizable,
        children: children2,
        onResize,
        label,
        ...rest
      } = child.props;
      panel = {
        children: children2,
        config: {
          minWidth: minWidth ?? 200,
          maxWidth: maxWidth ?? Infinity,
          defaultWidth: defaultWidth ?? 200,
          position: position ?? "start",
          resizable: resizable ?? true,
          onResize,
          label
        },
        domProps: rest,
        ref: "ref" in child ? child.ref : void 0
      };
    } else if (isContentSlot(child)) {
      if (content) {
        throw new Error("SplitLayout: only one Content slot is allowed");
      }
      const { children: children2, minWidth, ...rest } = child.props;
      content = {
        children: children2,
        config: {
          minWidth: minWidth ?? 250
        },
        domProps: rest,
        ref: "ref" in child ? child.ref : void 0
      };
    } else if (isResizeHandleSlot(child)) {
      if (resizeHandle) {
        throw new Error("SplitLayout: only one ResizeHandle slot is allowed");
      }
      const { children: children2, ...rest } = child.props;
      resizeHandle = {
        children: children2,
        domProps: rest,
        ref: "ref" in child ? child.ref : void 0
      };
    }
  });
  if (!panel) {
    throw new Error("Panel slot is required");
  }
  return { panel, content, resizeHandle };
}
