// @dynatrace/strato-components/layouts/split-layout/SplitLayout.sty.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SplitLayout_css_exports = {};
__export(SplitLayout_css_exports, {
  column: () => column,
  contentShellContainer: () => contentShellContainer,
  handleColumn: () => handleColumn,
  handleDecorations: () => handleDecorations,
  layout: () => layout,
  panelShellContainer: () => panelShellContainer,
  resizeHandle: () => resizeHandle
});
module.exports = __toCommonJS(SplitLayout_css_exports);
var import_AppRoot_css_ts_vanilla = require("../../core/components/app-root/AppRoot.css");
var import_SplitLayout_css_ts_vanilla = require("./SplitLayout.css");
var column = "_6q8dk21-3-6-0";
var contentShellContainer = "_6q8dk23-3-6-0 _6q8dk21-3-6-0";
var handleColumn = "_6q8dk24-3-6-0 _6q8dk21-3-6-0";
var handleDecorations = "_6q8dk25-3-6-0";
var layout = "_6q8dk20-3-6-0";
var panelShellContainer = "_6q8dk22-3-6-0 _6q8dk21-3-6-0";
var resizeHandle = "_6q8dk26-3-6-0";
