// @dynatrace/strato-components/layouts/split-layout/split-sizing-utils.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var split_sizing_utils_exports = {};
__export(split_sizing_utils_exports, {
  clampPanelWidth: () => clampPanelWidth,
  computeLayout: () => computeLayout
});
module.exports = __toCommonJS(split_sizing_utils_exports);
var import_width_utils = require("./width-utils.js");
function clampPanelWidth(split, containerWidth, config) {
  const { panelMin, panelMax, contentMin, handleWidth } = config;
  const minPx = (0, import_width_utils.resolveWidthPx)(panelMin, containerWidth, 0);
  const maxPx = (0, import_width_utils.resolveWidthPx)(panelMax, containerWidth, Infinity);
  const contentMinPx = (0, import_width_utils.resolveWidthPx)(contentMin, containerWidth, 0);
  let extrinsicMax = Infinity;
  if (containerWidth !== 0) {
    extrinsicMax = containerWidth - contentMinPx - handleWidth;
  }
  const finalMax = Math.max(minPx, Math.min(maxPx, extrinsicMax));
  return (0, import_width_utils.clamp)(minPx, split, finalMax);
}
function computeLayout(splitPx, containerWidth, config) {
  const { panelMin, panelMax, contentMin, handleWidth } = config;
  const panelWidth = clampPanelWidth(splitPx, containerWidth, config);
  const panelMinPx = (0, import_width_utils.resolveWidthPx)(panelMin, containerWidth, 0);
  const panelMaxPx = (0, import_width_utils.resolveWidthPx)(panelMax, containerWidth, Infinity);
  const contentMinPx = (0, import_width_utils.resolveWidthPx)(contentMin, containerWidth, 0);
  const contentMaxPx = Math.max(0, containerWidth - contentMinPx - handleWidth);
  const contentWidth = (0, import_width_utils.clamp)(
    contentMinPx,
    containerWidth - handleWidth - panelWidth,
    contentMaxPx
  );
  const contentWidthCss = `max(${(0, import_width_utils.toCssLength)(contentMin)}, 100% - ${panelWidth ? panelWidth + handleWidth : 0}px)`;
  return {
    panel: { width: panelWidth, minPx: panelMinPx, maxPx: panelMaxPx },
    content: {
      width: contentWidth,
      widthCss: contentWidthCss,
      minPx: contentMinPx,
      maxPx: contentMaxPx
    }
  };
}
