// @dynatrace/strato-components/layouts/split-layout/SplitLayout.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SplitLayout_exports = {};
__export(SplitLayout_exports, {
  SplitLayout: () => SplitLayout
});
module.exports = __toCommonJS(SplitLayout_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("@use-gesture/react");
var import_clsx = require("clsx");
var import_react2 = require("react");
var import_slots = require("./slots.js");
var styles = __toESM(require("./SplitLayout.sty.js"));
var import_useSplitSizing = require("./useSplitSizing.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useId = require("../../core/hooks/useId.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_useFocusRing = require("../../core/styles/useFocusRing.js");
const PanelShell = (0, import_react2.memo)(({ children }) => {
  return children;
});
const ContentShell = (0, import_react2.memo)(({ children }) => {
  return children;
});
const ResizeHandleShell = (0, import_react2.memo)(({ children }) => {
  return children;
});
const PanelColumn = (0, import_react2.forwardRef)(
  (props, forwardedRef) => {
    const { children, className, style, ...rest } = props;
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "div",
      {
        ...rest,
        ref: forwardedRef,
        className: (0, import_clsx.clsx)(styles.column, className),
        style,
        children
      }
    );
  }
);
const ContentColumn = (0, import_react2.forwardRef)(
  (props, forwardedRef) => {
    const { children, className, style, ...rest } = props;
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "div",
      {
        ...rest,
        ref: forwardedRef,
        className: (0, import_clsx.clsx)(styles.column, className),
        style,
        children
      }
    );
  }
);
const ResizeHandle = (0, import_react2.forwardRef)((props, forwardedRef) => {
  const { onDrag, onDragStart, onDragEnd, panelId, ariaLabel, ariaLabelledBy } = (0, import_react2.useContext)(ActionsContext);
  const ref = (0, import_react2.useRef)(null);
  const mergedRefs = (0, import_useMergeRefs.useMergeRefs)([ref, forwardedRef]);
  const { focusProps, focusClassName } = (0, import_useFocusRing.useFocusRing)({
    focusWithin: true,
    ignoreModality: true
  });
  (0, import_react.useGesture)(
    {
      onDragStart,
      onDragEnd,
      onDrag: (state) => {
        onDrag(state.movement[0]);
      }
    },
    { target: ref }
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      ...props,
      ...focusProps,
      role: "separator",
      ref: mergedRefs,
      className: (0, import_clsx.clsx)(styles.resizeHandle, focusClassName),
      tabIndex: 0,
      "aria-controls": panelId,
      "aria-label": ariaLabelledBy ? void 0 : ariaLabel,
      "aria-labelledby": ariaLabelledBy
    }
  );
});
const ActionsContext = (0, import_react2.createContext)({
  onDragStart: () => void 0,
  onDrag: () => void 0,
  onDragEnd: () => void 0,
  panelId: "",
  ariaLabel: void 0,
  ariaLabelledBy: void 0
});
const SplitLayoutRoot = (0, import_react2.memo)(
  (0, import_react2.forwardRef)((props, forwardedRef) => {
    const { children, className: rootClassname, ...rootProps } = props;
    const [behavioralTrackingProps, restRootProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(rootProps);
    const slots = (0, import_slots.extractSlotsProps)(children);
    const rootRef = (0, import_react2.useRef)(null);
    const rootMergedRef = (0, import_useMergeRefs.useMergeRefs)([forwardedRef, rootRef]);
    const generatedPanelId = (0, import_useId.useId)();
    const panelId = slots.panel.domProps.id ?? generatedPanelId;
    const panelRef = (0, import_react2.useRef)(null);
    const contentRef = (0, import_react2.useRef)(null);
    const panelMergedRef = (0, import_useMergeRefs.useMergeRefs)([panelRef, slots.panel.ref]);
    const contentMergedRef = (0, import_useMergeRefs.useMergeRefs)([contentRef, slots.content?.ref]);
    const sizing = (0, import_useSplitSizing.useSplitSizing)({
      panel: {
        min: slots.panel.config.minWidth,
        max: slots.panel.config.maxWidth,
        default: slots.panel.config.defaultWidth,
        position: slots.panel.config.position,
        resizable: slots.panel.config.resizable,
        onResize: slots.panel.config.onResize
      },
      content: {
        min: slots.content?.config.minWidth ?? 0
      },
      refs: {
        root: rootRef,
        panel: panelRef,
        content: contentRef
      }
    });
    const isDefaultPosition = slots.panel.config.position === "start";
    const activeSizing = isDefaultPosition ? sizing.panel : sizing.content;
    const panelColumn = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      PanelColumn,
      {
        ...slots.panel.domProps,
        id: panelId,
        style: { ...slots.panel.domProps.style, width: sizing.panel.widthCss },
        ref: panelMergedRef,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(PanelShell, { children: slots.panel.children })
      }
    );
    const contentColumn = slots.content ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      ContentColumn,
      {
        ...slots.content.domProps,
        style: {
          ...slots.content.domProps.style,
          width: sizing.content.widthCss,
          // content needs an explicit minWidth when it is the leading column
          // (end position) to prevent it from collapsing before the resize handle
          minWidth: isDefaultPosition ? void 0 : slots.content.config.minWidth
        },
        ref: contentMergedRef,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(ContentShell, { children: slots.content.children })
      }
    ) : null;
    const resizeHandleDecorations = slots.resizeHandle?.children ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: styles.handleDecorations, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(ResizeHandleShell, { children: slots.resizeHandle.children }) }) : null;
    const resizeHandle = slots.panel.config.resizable ? /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      "div",
      {
        ...slots.resizeHandle?.domProps,
        className: (0, import_clsx.clsx)(
          styles.handleColumn,
          slots.resizeHandle?.domProps?.className
        ),
        ref: slots.resizeHandle?.ref,
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            ResizeHandle,
            {
              "aria-valuemin": activeSizing.minPx,
              "aria-valuemax": isFinite(activeSizing.maxPx) ? activeSizing.maxPx : void 0,
              "aria-valuenow": activeSizing.width,
              "aria-valuetext": `${activeSizing.width}px`,
              "aria-orientation": "vertical"
            }
          ),
          resizeHandleDecorations
        ]
      }
    ) : null;
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "div",
      {
        ...restRootProps,
        ...behavioralTrackingProps,
        className: (0, import_clsx.clsx)(styles.layout, rootClassname),
        ref: rootMergedRef,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
          ActionsContext.Provider,
          {
            value: {
              ...sizing.dragActions,
              panelId,
              ariaLabel: slots.resizeHandle?.domProps["aria-label"] ?? slots.panel.config.label,
              ariaLabelledBy: slots.resizeHandle?.domProps["aria-labelledby"]
            },
            children: [
              isDefaultPosition ? panelColumn : contentColumn,
              resizeHandle,
              isDefaultPosition ? contentColumn : panelColumn
            ]
          }
        )
      }
    );
  })
);
SplitLayoutRoot.displayName = "SplitLayout.Root";
const SplitLayout = Object.assign(SplitLayoutRoot, {
  Panel: import_slots.PanelSlot,
  Content: import_slots.ContentSlot,
  ResizeHandle: import_slots.ResizeHandleSlot
});
