// @dynatrace/strato-components/layouts/split-layout/useSplitSizing.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useSplitSizing_exports = {};
__export(useSplitSizing_exports, {
  useSplitSizing: () => useSplitSizing
});
module.exports = __toCommonJS(useSplitSizing_exports);
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_use_resize_observer = __toESM(require("use-resize-observer"));
var import_split_sizing_utils = require("./split-sizing-utils.js");
var import_width_utils = require("./width-utils.js");
function useSplitSizing(options) {
  const { panel, content, refs } = options;
  const [containerWidth, setContainerWidth] = (0, import_react.useState)(0);
  const [splitPx, setSplitPx] = (0, import_react.useState)(() => {
    if (typeof panel.default === "number") {
      return panel.default;
    }
    return -1;
  });
  const splitPxRef = (0, import_react.useRef)(splitPx);
  const isDraggingRef = (0, import_react.useRef)(false);
  const hasBeenManuallyResizedRef = (0, import_react.useRef)(false);
  const dragStartPx = (0, import_react.useRef)(null);
  if (!isDraggingRef.current) {
    splitPxRef.current = splitPx;
  }
  const handleWidth = panel.resizable ? 8 : 0;
  const contentMinCss = (0, import_width_utils.toCssLength)(content.min);
  const sizingConfig = (0, import_react.useMemo)(
    () => ({
      panelMin: panel.min,
      panelMax: panel.max,
      contentMin: content.min,
      handleWidth
    }),
    [panel.min, panel.max, content.min, handleWidth]
  );
  const getClampedSplit = (0, import_react.useCallback)(
    (split, container) => (0, import_split_sizing_utils.clampPanelWidth)(split, container, sizingConfig),
    [sizingConfig]
  );
  (0, import_react.useLayoutEffect)(
    function initializeSplit() {
      if (containerWidth === 0) {
        return;
      }
      if (typeof panel.default === "number") {
        if (splitPx !== -1) {
          return;
        }
        setSplitPx(panel.default);
        return;
      }
      if (hasBeenManuallyResizedRef.current) {
        return;
      }
      setSplitPx((0, import_width_utils.convertToPx)(panel.default, containerWidth));
    },
    [containerWidth, panel.default, splitPx]
  );
  const KEEP_SQUISHED_SIZES = false;
  const throttledSync = (0, import_react.useMemo)(
    () => (0, import_lodash_es.throttle)((width) => {
      const expected = getClampedSplit(splitPxRef.current, width);
      if (splitPxRef.current !== expected) {
        setSplitPx(expected);
      }
    }, 100),
    [getClampedSplit]
  );
  (0, import_use_resize_observer.default)({
    ref: refs.root,
    onResize: (size) => {
      if (!size.width) {
        return;
      }
      setContainerWidth(size.width);
      if (KEEP_SQUISHED_SIZES) {
        throttledSync(size.width);
      }
    }
  });
  const { position: panelPosition, onResize: onPanelResize } = panel;
  const dragActions = (0, import_react.useMemo)(
    () => ({
      onDragStart: () => {
        isDraggingRef.current = true;
        const containerW = refs.root.current?.offsetWidth ?? 0;
        const clamped = getClampedSplit(splitPxRef.current, containerW);
        splitPxRef.current = clamped;
        setSplitPx(clamped);
        dragStartPx.current = clamped;
      },
      onDrag: (mx) => {
        const containerW = refs.root.current?.offsetWidth ?? 0;
        if (dragStartPx.current === null || containerW === 0) {
          return;
        }
        const sign = panelPosition === "start" ? 1 : -1;
        const next = dragStartPx.current + mx * sign;
        const clamped = getClampedSplit(next, containerW);
        splitPxRef.current = clamped;
        if (refs.panel.current) {
          refs.panel.current.style.width = `${clamped}px`;
        }
        if (refs.content.current) {
          refs.content.current.style.width = `max(${contentMinCss}, 100% - ${clamped + handleWidth}px)`;
        }
      },
      onDragEnd: () => {
        isDraggingRef.current = false;
        hasBeenManuallyResizedRef.current = true;
        dragStartPx.current = null;
        const rootWidth = refs.root.current?.offsetWidth ?? 0;
        const clamped = getClampedSplit(splitPxRef.current, rootWidth);
        splitPxRef.current = clamped;
        setSplitPx(clamped);
        onPanelResize?.(clamped);
      }
    }),
    [
      getClampedSplit,
      panelPosition,
      onPanelResize,
      handleWidth,
      contentMinCss,
      refs
    ]
  );
  const layout = (0, import_split_sizing_utils.computeLayout)(splitPx, containerWidth, sizingConfig);
  const isUnresolved = splitPx === -1;
  const panelWidthCss = isUnresolved ? (0, import_width_utils.toCssLength)(panel.default) : `${layout.panel.width}px`;
  const contentWidthCss = isUnresolved ? `max(${(0, import_width_utils.toCssLength)(content.min)}, 100% - ${(0, import_width_utils.toCssLength)(
    panel.default
  )} - ${handleWidth}px)` : layout.content.widthCss;
  return {
    ...layout,
    panel: { ...layout.panel, widthCss: panelWidthCss },
    content: { ...layout.content, widthCss: contentWidthCss },
    handleWidth,
    dragActions
  };
}
