// @dynatrace/strato-components/esm/core/utils/modality.js
import { createStore } from "../store/store.js";
const modalityStore = createStore("unknown");
const ignoreMap = [
  "Alt",
  "AltGraph",
  "CapsLock",
  "Control",
  "FnLock",
  "Fn",
  "Hyper",
  "Meta",
  // Command (Mac), in some browsers also Windows-key
  "NumLock",
  "OS",
  // Windows-key
  "ScrollLock",
  "Super",
  // Virtual keyboard key
  "Symbol",
  "SymbolLock"
];
const eventTypeMap = {
  keydown: "keyboard",
  keyup: "keyboard",
  mousedown: "mouse",
  mousemove: "mouse",
  MSPointerDown: "pointer",
  MSPointerMove: "pointer",
  pointerdown: "pointer",
  pointermove: "pointer",
  touchstart: "touch",
  touchend: "touch",
  wheel: "wheel"
};
const navigationalKeys = [
  "Tab",
  "Esc",
  "Escape",
  " ",
  "ArrowDown",
  "Down",
  "ArrowUp",
  "Up",
  "ArrowLeft",
  "Left",
  "ArrowRight",
  "Right",
  "End",
  "Home",
  "PageDown",
  "PageUp"
];
function handleKeyboardInput(event) {
  if (ignoreMap.includes(event.key)) {
    modalityStore.setState("pointer");
    return;
  }
  if (!event.target || navigationalKeys.includes(event.key)) {
    modalityStore.setState("keyboard");
  }
}
function handlePointerInput(event) {
  const modality = eventTypeMap[event.type];
  switch (modality) {
    case "touch":
    case "pen":
      modalityStore.setState("touch");
      break;
    case "mouse":
    case "pointer":
    case "wheel":
      modalityStore.setState("pointer");
      break;
  }
}
let isInitialized = false;
function initModalityDetection() {
  if (isInitialized) {
    return;
  }
  window.addEventListener("pointerdown", handlePointerInput);
  window.addEventListener("keydown", handleKeyboardInput, { capture: true });
  window.addEventListener("wheel", handlePointerInput);
  isInitialized = true;
}
export {
  initModalityDetection,
  modalityStore
};
