// @dynatrace/strato-components/esm/tables/DataTable/features/ColumnFontStyle/ColumnFontStyle.js
import { functionalUpdate } from "@tanstack/react-table";
import { useControllableState } from "../../../../core/hooks/useControllableState.js";
import { useDataTableConfigElement } from "../../../hooks/useDataTableConfig.js";
const DataTableColumnFontStyle = {
  createColumn(column, table) {
    column.toggleFontStyle = (value) => {
      table.setColumnFontStyle((old) => {
        let newValue = value;
        if (newValue === void 0) {
          if (old[column.id] !== void 0) {
            newValue = old[column.id] === "code" ? "text" : "code";
          } else {
            newValue = table.getVisualVariantOptions().fontStyle === "code" ? "text" : "code";
          }
        }
        return {
          ...old,
          [column.id]: newValue
        };
      });
    };
    column.getFontStyle = () => {
      const fontStyleState = table.getState().columnFontStyle;
      if (fontStyleState?.[column.id]) {
        return fontStyleState?.[column.id];
      }
      return column.columnDef.meta?.fontStyle ?? table.getVisualVariantOptions().fontStyle;
    };
  },
  createTable(table) {
    table.setColumnFontStyle = (updater) => {
      const newValue = functionalUpdate(
        updater,
        table.getState().columnFontStyle ?? {}
      );
      table.options.onColumnFontStyleChange?.(newValue);
    };
  }
};
function useColumnFontStyle(props, options) {
  const configFontStyle = useDataTableConfigElement("fontStyle");
  const [columnFontStyle, onColumnFontStyleChange] = useControllableState({
    value: props.fontStyle ?? configFontStyle,
    defaultValue: props.defaultFontStyle,
    onChange: props.onFontStyleChange
  });
  options.state = { ...options.state, columnFontStyle };
  options.onColumnFontStyleChange = onColumnFontStyleChange;
}
export {
  DataTableColumnFontStyle,
  useColumnFontStyle
};
