// @dynatrace/strato-components/tables/DataTable/features/CellFormatter/CellFormatter.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CellFormatter_exports = {};
__export(CellFormatter_exports, {
  applyColumnFormatters: () => applyColumnFormatters,
  formatCellValue: () => formatCellValue
});
module.exports = __toCommonJS(CellFormatter_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_units = require("@dynatrace-sdk/units");
var import_default_formatters = require("../../../../core/utils/default-formatters.js");
var import_logging = require("../../../../core/utils/logging.js");
var import_string_utils = require("../../../../core/utils/string-utils.js");
const DATETIME_FORMAT_OPTIONS = {
  // e.g. "12/21/2024" for US English locale
  date: {
    year: "numeric",
    month: "numeric",
    day: "numeric"
  },
  // e.g. "5:45:30 PM" for US English locale
  time: {
    hour: "numeric",
    minute: "numeric",
    second: "numeric"
  },
  // e.g. "12/21/2024, 5:45:30 PM" for US English locale
  datetime: import_default_formatters.DATETIME_DEFAULT_FORMAT_OPTIONS
};
function isDatetimeShorthand(formatter) {
  return formatter === "date" || formatter === "time" || formatter === "datetime";
}
function resolveDatetimeFormatter(formatter) {
  if (!formatter) {
    return void 0;
  }
  if (isDatetimeShorthand(formatter)) {
    return DATETIME_FORMAT_OPTIONS[formatter];
  }
  return formatter;
}
const ColumnCellDatetimeFormatMemo = (0, import_react.memo)(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  (props) => {
    return props.value ? (0, import_units.formatDate)(Number(props.value), props.formatter) : "";
  }
);
const ColumnCellCurrencyFormatMemo = (0, import_react.memo)(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  (props) => {
    if (props.value === null || props.value === void 0 || isNaN(props.value)) {
      return "";
    }
    const { currency, ...options } = props.formatter ?? {};
    if (!currency) {
      return (0, import_default_formatters.formatCurrencyDefault)(Number(props.value), options);
    }
    return (0, import_units.formatCurrency)(Number(props.value), currency, options);
  }
);
const ColumnCellFormatMemo = (0, import_react.memo)(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  (props) => {
    return (0, import_units.format)(Number(props.value), props.formatter);
  }
);
const ColumnCellFormatLongMemo = (0, import_react.memo)(
  (props) => {
    return (0, import_units.formatLong)(props.value, props.formatter);
  }
);
function formatColumnCell(columnDef) {
  if (columnDef.columnType === "sparkline" || columnDef.columnType === "meterbar" || columnDef.columnType === "gantt" || columnDef.columnType === "log-content" || columnDef.columnType === "markdown") {
    return;
  }
  if (columnDef.columnType === "date" || columnDef.columnType === "datetime") {
    const resolvedFormatter = resolveDatetimeFormatter(columnDef.formatter);
    return ({ getValue }) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      ColumnCellDatetimeFormatMemo,
      {
        value: getValue(),
        formatter: resolvedFormatter
      }
    );
  }
  if (columnDef.columnType === "currency") {
    return ({ getValue }) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      ColumnCellCurrencyFormatMemo,
      {
        value: getValue(),
        formatter: columnDef.formatter
      }
    );
  }
  if (columnDef.columnType === "long") {
    return ({ getValue }) => {
      const value = getValue();
      if (typeof value === "string" || typeof value === "bigint" || typeof value === "number") {
        return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          ColumnCellFormatLongMemo,
          {
            value,
            formatter: columnDef.formatter
          }
        );
      }
      return String(value);
    };
  }
  return ({ getValue }) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(ColumnCellFormatMemo, { value: getValue(), formatter: columnDef.formatter });
}
function applyColumnFormatters(columnDef) {
  if (columnDef.columnType === "sparkline" || columnDef.columnType === "meterbar" || columnDef.columnType === "gantt" || columnDef.columnType === "log-content" || columnDef.columnType === "markdown") {
    return {};
  }
  if (columnDef.formatter) {
    return {
      cell: formatColumnCell(columnDef)
    };
  }
  return {};
}
function formatCellValue(value, column) {
  if (column.columnType === "sparkline" || column.columnType === "meterbar" || column.columnType === "gantt" || column.columnType === "log-content" || column.columnType === "markdown") {
    return "";
  }
  if (value === null || value === void 0) {
    (0, import_logging.devLog)(`The value '${value}' can not be formatted!`);
    return "";
  }
  const cleanedValue = typeof value === "number" || value instanceof Date ? value : Number.isNaN(+value) ? null : (0, import_string_utils.convertToNumberIfPureNumber)(value);
  if (column.formatter && // DEPRECATION: [APPDEV-16044] BREAKING-CHANGE: Remove 'date' columnType.
  column.columnType !== "date" && column.columnType !== "datetime" && column.columnType !== "currency" && typeof cleanedValue === "number") {
    const formatOptions = column.formatter;
    return (0, import_units.format)(cleanedValue, formatOptions);
  }
  if (column.columnType === "bit" && typeof cleanedValue === "number") {
    const formatOptions = { input: import_units.units.data.bit };
    return (0, import_units.format)(cleanedValue, formatOptions);
  }
  if (column.columnType === "number" && typeof cleanedValue === "number") {
    const formatOptions = column.formatter;
    if (formatOptions) {
      return (0, import_units.format)(cleanedValue, formatOptions);
    }
    return (0, import_units.adjustFractionDigits)(cleanedValue);
  }
  if (column.columnType === "long" && (typeof value === "string" || typeof value === "bigint" || typeof value === "number")) {
    return (0, import_units.formatLong)(value, column.formatter);
  }
  if (
    // DEPRECATION: [APPDEV-16044] BREAKING-CHANGE: Remove 'date' columnType
    (column.columnType === "date" || column.columnType === "datetime") && (typeof cleanedValue === "number" || cleanedValue instanceof Date)
  ) {
    const resolvedFormatter = resolveDatetimeFormatter(column.formatter);
    if (resolvedFormatter) {
      return (0, import_units.formatDate)(cleanedValue, resolvedFormatter);
    } else {
      return (0, import_default_formatters.formatDatetimeDefault)(cleanedValue);
    }
  }
  if (column.columnType === "currency" && typeof cleanedValue === "number") {
    const formatOptions = column.formatter;
    const { currency, ...remainingOptions } = formatOptions ?? {};
    if (!currency) {
      return (0, import_default_formatters.formatCurrencyDefault)(cleanedValue, remainingOptions);
    }
    return (0, import_units.formatCurrency)(cleanedValue, currency, remainingOptions);
  }
  (0, import_logging.devLog)(`The value '${value}' can not be formatted!`);
  return String(value);
}
