// @dynatrace/strato-components/filters/segment-selector/parts/segmentsPresets/SegmentsPresets.sty.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SegmentsPresets_css_exports = {};
__export(SegmentsPresets_css_exports, {
  and: () => and,
  buttonLabel: () => buttonLabel,
  inputGroup: () => inputGroup,
  inputGroupSelected: () => inputGroupSelected,
  list: () => list,
  listItem: () => listItem,
  presetButton: () => presetButton,
  sectionHeadline: () => sectionHeadline,
  selectionMark: () => selectionMark,
  suffixButton: () => suffixButton
});
module.exports = __toCommonJS(SegmentsPresets_css_exports);
var import_AppRoot_css_ts_vanilla = require("../../../../core/components/app-root/AppRoot.css");
var import_SegmentsPresets_css_ts_vanilla = require("./SegmentsPresets.css");
var and = "_1oehqvw5-3-6-0";
var buttonLabel = "_1oehqvw6-3-6-0";
var inputGroup = "_1oehqvw9-3-6-0";
var inputGroupSelected = "_1oehqvwa-3-6-0";
var list = "_1oehqvw3-3-6-0";
var listItem = "_1oehqvw4-3-6-0";
var presetButton = "_1oehqvwb-3-6-0";
var sectionHeadline = "_1oehqvw7-3-6-0";
var selectionMark = "_1oehqvw8-3-6-0";
var suffixButton = "_1oehqvwc-3-6-0";
