// @dynatrace/strato-components/filters/segment-selector/parts/segmentsPresets/SegmentsPreset.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SegmentsPreset_exports = {};
__export(SegmentsPreset_exports, {
  SegmentsPreset: () => SegmentsPreset,
  expandAndCollapseAnimationVariants: () => expandAndCollapseAnimationVariants
});
module.exports = __toCommonJS(SegmentsPreset_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_query = require("@tanstack/react-query");
var import_clsx = __toESM(require("clsx"));
var import_framer_motion = require("framer-motion");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_animations_js = __toESM(require("@dynatrace/strato-design-tokens/animations-js"));
var import_strato_icons = require("@dynatrace/strato-icons");
var import_app_environment = require("@dynatrace-sdk/app-environment");
var import_segments_presets = require("@dynatrace-sdk/segments-presets");
var styles = __toESM(require("./SegmentsPresets.sty.js"));
var import_Button = require("../../../../buttons/button/Button.js");
var import_useOverlayStacking = require("../../../../core/hooks/useOverlayStacking.js");
var import_framer_motion_easings = require("../../../../core/utils/framer-motion-easings.js");
var import_Flex = require("../../../../layouts/flex/Flex.js");
var import_InputGroup = require("../../../../layouts/input-group/InputGroup.js");
var import_Tooltip = require("../../../../overlays/tooltip/Tooltip.js");
var import_textStyle_css = require("../../../../styles/textStyle.sty.js");
var import_Text = require("../../../../typography/text/Text.js");
var import_queryHelper = require("../../../helper/queryHelper.js");
var import_SegmentsCache = require("../../contexts/SegmentsCache/SegmentsCache.js");
var import_SegmentsUIContext = require("../../contexts/SegmentsUI/SegmentsUIContext.js");
var import_SegmentsUIDispatchContext = require("../../contexts/SegmentsUI/SegmentsUIDispatchContext.js");
var import_segmentValidationHelper = require("../../contexts/SegmentsUI/segmentValidationHelper.js");
var import_useSegments = require("../../hooks/useSegments.js");
var import_utils = require("../../utils.js");
var animationHelperStyles = __toESM(require("../Overlay.sty.js"));
const expandAndCollapseAnimationVariants = {
  enter: {
    height: "auto",
    opacity: 1,
    translateY: 0,
    transition: {
      duration: import_animations_js.default.ExpandCollapse.Duration.Base,
      ease: import_framer_motion_easings.expandCollapseEasing
    }
  },
  exit: {
    height: 0,
    opacity: 0,
    translateY: -10,
    transition: {
      duration: import_animations_js.default.ExpandCollapse.Duration.Base,
      ease: import_framer_motion_easings.expandCollapseEasing
    }
  }
};
const SegmentsPresetText = ({
  segment,
  isFirst
}) => {
  const andText = /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Text.Text, { textStyle: "base", className: styles.and, children: [
    "\xA0",
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_react_intl.FormattedMessage,
      {
        defaultMessage: "and",
        description: "Text that is separating between multiple Segments in one preset.",
        id: "GeQL1Mbr4RzmlFJk"
      }
    ),
    "\xA0"
  ] });
  const connectingString = isFirst ? "" : andText;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_react.Fragment, { children: [
    connectingString,
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(SegmentText, { segment })
  ] }, segment.id);
};
function usePinningTooltipText(pinLimitReached, isPinned) {
  const intl = (0, import_react_intl.useIntl)();
  if (pinLimitReached && !isPinned) {
    return intl.formatMessage({
      defaultMessage: "You can't pin more than 10 items.",
      id: "Y/6wwuz1ueL+7V2j",
      description: "Tooltip text for pinning limit."
    });
  }
  if (!isPinned) {
    return intl.formatMessage(
      {
        defaultMessage: "Pin in {appName}",
        id: "RPdRdtDJumoOO3os",
        description: "Tooltip text for pin button."
      },
      { appName: (0, import_app_environment.getAppName)() }
    );
  }
  return intl.formatMessage({
    defaultMessage: "Unpin",
    id: "uCbwlgMp2BNGzNro",
    description: "Text for the unpin button of a segment preset."
  });
}
function updateCachePinMutation(old, id, segments) {
  const pinnedPresets = old.pinnedPresets;
  pinnedPresets.unshift({ id, segments });
  return {
    ...old,
    pinnedPresets
  };
}
function updateCacheUnpinMutation(old, id) {
  const pinnedPresets = old.pinnedPresets;
  const pinnedPresetIndex = pinnedPresets.findIndex(
    (preset) => preset.id === id
  );
  if (pinnedPresetIndex === -1) {
    return old;
  }
  pinnedPresets.splice(pinnedPresetIndex, 1);
  return {
    recentPresets: old.recentPresets,
    pinnedPresets
  };
}
const PresetTooltipText = ({ segments }) => {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Flex.Flex, { flexDirection: "column", gap: 4, children: segments.map((segment) => {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      SegmentText,
      {
        segment,
        variableClassName: (0, import_textStyle_css.textStyle)({
          fontStyle: "text",
          textStyle: "small"
        })
      },
      segment.id
    );
  }) });
};
const SegmentText = ({
  segment,
  variableClassName
}) => {
  const { getSegment } = (0, import_SegmentsCache.useSegmentsCache)();
  const intl = (0, import_react_intl.useIntl)();
  const { data } = (0, import_react_query.useQuery)(
    {
      queryKey: ["getSegment", segment.id],
      queryFn: () => getSegment(segment.id),
      notifyOnChangeProps: ["dataUpdatedAt"]
    },
    import_SegmentsCache.segmentsClient.queryClient
  );
  let variables = "";
  if (segment.variables && segment.variables.length > 0) {
    variables = `: ${segment.variables[0].values.map((variable) => (0, import_utils.getReadableValue)(variable, intl)).join(", ")}`;
  }
  if (variableClassName) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("span", { children: [
      data?.segment.name ?? segment.id,
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { className: variableClassName, children: variables })
    ] });
  }
  return `${data?.segment.name ?? segment.id}${variables}`;
};
const SegmentsPreset = (props) => {
  const shouldReduceMotion = (0, import_framer_motion.useReducedMotion)();
  const {
    uiPreset,
    id,
    isPinned,
    isSelected,
    pinLimitReached
  } = props;
  const { segments } = uiPreset;
  const [isAnimating, setIsAnimating] = (0, import_react.useState)(false);
  const { setUISegments } = (0, import_SegmentsUIContext.useSegmentsUIContext)();
  const { overlayState } = (0, import_react.useContext)(import_SegmentsUIDispatchContext.UISegmentsDispatchContext);
  const { setSegments } = (0, import_useSegments.useSegments)();
  const pin = (0, import_react_query.useMutation)(
    {
      mutationKey: ["pin-segment-preset", id],
      mutationFn: import_segments_presets.pinSegmentsPreset,
      onMutate: async () => (0, import_queryHelper.runOptimisticMutation)(
        "getSegmentsPresets",
        (prevState) => updateCachePinMutation(prevState, id, segments),
        [
          "pin-segment-preset",
          "unpin-segment-preset",
          id,
          "getSegmentsPresets"
        ]
      ),
      onError: (error, _, context) => (0, import_queryHelper.handleFailedOptimisticMutation)("getSegmentsPresets", error, context),
      onSettled: () => (0, import_queryHelper.handleSettledOptimisticMutation)(
        ["getSegmentsPresets"],
        ["pin-segment-preset", "unpin-segment-preset", id]
      )
    },
    import_SegmentsCache.segmentsClient.queryClient
  );
  const unpin = (0, import_react_query.useMutation)(
    {
      mutationKey: ["unpin-segment-preset", id],
      mutationFn: import_segments_presets.unpinSegmentsPreset,
      onMutate: async () => (0, import_queryHelper.runOptimisticMutation)(
        "getSegmentsPresets",
        (prevState) => updateCacheUnpinMutation(prevState, id),
        [
          "pin-segment-preset",
          "unpin-segment-preset",
          id,
          "getSegmentsPresets"
        ]
      ),
      onError: (error, _, context) => (0, import_queryHelper.handleFailedOptimisticMutation)("getSegmentsPresets", error, context),
      onSettled: () => (0, import_queryHelper.handleSettledOptimisticMutation)(
        ["getSegmentsPresets"],
        ["pin-segment-preset", "unpin-segment-preset", id]
      )
    },
    import_SegmentsCache.segmentsClient.queryClient
  );
  const applyPreset = (0, import_react_query.useMutation)(
    {
      mutationKey: ["apply-segment-preset", id],
      mutationFn: () => {
        const flattenedSegments = segments.map((segment) => {
          if (segment.variables) {
            const flattenedVariables = segment.variables.map((variable) => {
              const values = (0, import_utils.flattenQueryResultToValues)([], variable.values);
              return {
                ...variable,
                values
              };
            });
            return { ...segment, variables: flattenedVariables };
          }
          return segment;
        });
        setUISegments(flattenedSegments, true);
        return setSegments(flattenedSegments);
      },
      onSuccess: async () => {
        const isValid = await (0, import_segmentValidationHelper.areUiSegmentsValid)(segments);
        if (isValid) {
          overlayState?.close();
        }
      }
    },
    import_SegmentsCache.segmentsClient.queryClient
  );
  const color = isSelected ? "primary" : "neutral";
  const pinningTooltipText = usePinningTooltipText(pinLimitReached, isPinned);
  const icon = isPinned ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.PinOffIcon, {}) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.PinIcon, {});
  const pinButtonDataDtProperties = `state:${isPinned ? "unpin" : "pin"}`;
  const onClick = () => isPinned ? unpin.mutate(uiPreset.id) : pin.mutate(segments);
  const labelRef = (0, import_react.useRef)(null);
  const [showTooltip, setShowTooltip] = (0, import_react.useState)(false);
  const stackingContext = (0, import_useOverlayStacking.useOverlayStacking)(void 0, "overlay");
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_framer_motion.motion.div,
    {
      variants: shouldReduceMotion ? {} : expandAndCollapseAnimationVariants,
      initial: "exit",
      animate: "enter",
      exit: "exit",
      className: (0, import_clsx.default)({
        [animationHelperStyles.isAnimating]: isAnimating
      }),
      onAnimationStart: () => setIsAnimating(true),
      onAnimationComplete: () => setIsAnimating(false),
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
        import_InputGroup.InputGroup,
        {
          className: (0, import_clsx.default)([
            styles.inputGroup,
            { [styles.inputGroupSelected]: isSelected }
          ]),
          children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_Tooltip.Tooltip,
              {
                placement: "left",
                fallbackPlacements: ["top", "bottom", "right"],
                text: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(PresetTooltipText, { segments }),
                open: showTooltip,
                onOpenChange: (isOpen) => {
                  const hasOverFlow = (labelRef.current?.scrollWidth ?? 0) > (labelRef.current?.clientWidth ?? 0);
                  setShowTooltip(isOpen && hasOverFlow);
                },
                className: (0, import_textStyle_css.textStyle)({
                  fontStyle: "text",
                  textStyle: "small-emphasized"
                }),
                style: { zIndex: Number(stackingContext.style.zIndex) + 1 },
                children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  import_Button.Button,
                  {
                    color,
                    width: "100%",
                    textAlign: "start",
                    className: (0, import_clsx.default)(styles.presetButton, {
                      [styles.selectionMark]: isSelected
                    }),
                    onClick: () => applyPreset.mutate(),
                    children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Label, { className: styles.buttonLabel, ref: labelRef, children: segments.map((segment, index) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                      SegmentsPresetText,
                      {
                        segment,
                        isFirst: index === 0
                      },
                      segment.id
                    )) })
                  }
                )
              }
            ),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_Tooltip.Tooltip,
              {
                placement: "right",
                text: pinningTooltipText,
                className: (0, import_textStyle_css.textStyle)({
                  fontStyle: "text",
                  textStyle: "small-emphasized"
                }),
                style: { zIndex: Number(stackingContext.style.zIndex) + 1 },
                children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  import_Button.Button,
                  {
                    color,
                    className: styles.suffixButton,
                    onClick,
                    disabled: pinLimitReached && !isPinned,
                    "aria-label": pinningTooltipText,
                    "data-dt-name": "PinUnpin",
                    "data-dt-properties": pinButtonDataDtProperties,
                    children: icon
                  }
                )
              }
            )
          ]
        }
      )
    }
  );
};
