// @dynatrace/strato-components/filters/segment-selector/parts/segmentsPresets/SegmentsPresetsList.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SegmentsPresetsList_exports = {};
__export(SegmentsPresetsList_exports, {
  SegmentsPresetList: () => SegmentsPresetList
});
module.exports = __toCommonJS(SegmentsPresetsList_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_query = require("@tanstack/react-query");
var import_clsx = __toESM(require("clsx"));
var import_framer_motion = require("framer-motion");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_spacings = __toESM(require("@dynatrace/strato-design-tokens/spacings"));
var import_app_environment = require("@dynatrace-sdk/app-environment");
var import_segments_presets = require("@dynatrace-sdk/segments-presets");
var import_isSelectedPreset = require("./isSelectedPreset.js");
var import_SegmentsPreset = require("./SegmentsPreset.js");
var styles = __toESM(require("./SegmentsPresets.sty.js"));
var import_Divider = require("../../../../layouts/divider/Divider.js");
var import_Flex = require("../../../../layouts/flex/Flex.js");
var import_Text = require("../../../../typography/text/Text.js");
var import_SegmentsCache = require("../../contexts/SegmentsCache/SegmentsCache.js");
var import_useSegments = require("../../hooks/useSegments.js");
var animationHelperStyles = __toESM(require("../Overlay.sty.js"));
var import_ShowMorePanel = require("../ShowMorePanel/ShowMorePanel.js");
const PIN_LIMIT = 10;
const SegmentsPresetList = ({
  showDivider
}) => {
  const { segments } = (0, import_useSegments.useSegments)();
  const { data, isLoading } = (0, import_react_query.useQuery)(
    {
      queryKey: ["getSegmentsPresets"],
      queryFn: import_segments_presets.getSegmentsPresets,
      notifyOnChangeProps: ["dataUpdatedAt"]
    },
    import_SegmentsCache.segmentsClient.queryClient
  );
  const [isAnimating, setIsAnimating] = (0, import_react.useState)(false);
  if (!data || data.pinnedPresets.length === 0 && data.recentPresets.length === 0 || isLoading) {
    return null;
  }
  const pinLimitReached = data.pinnedPresets.length >= PIN_LIMIT;
  const recentPresets = data.recentPresets?.map((preset) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "li",
    {
      className: styles.listItem,
      "data-dt-name": "Recent",
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_SegmentsPreset.SegmentsPreset,
        {
          uiPreset: preset,
          id: preset.id,
          pinLimitReached,
          isPinned: data.pinnedPresets.some(
            (pinnedPreset) => preset.id === pinnedPreset.id
          ),
          isSelected: (0, import_isSelectedPreset.isSelectedPreset)(segments, preset.segments)
        }
      )
    },
    `recent-${preset.id}`
  ));
  const primaryContent = recentPresets.length > 3 ? recentPresets.slice(0, 2) : recentPresets;
  const extendedContent = recentPresets.length > 3 ? recentPresets.slice(2) : null;
  const showRecent = data.recentPresets.length > 0;
  const showPinned = data.pinnedPresets.length > 0;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    showDivider && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Divider.Divider, {}),
    /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      import_Flex.Flex,
      {
        flexDirection: "column",
        width: "100%",
        px: 0,
        pb: 4,
        gap: 0,
        pt: showDivider ? 12 : 0,
        className: (0, import_clsx.default)({
          [animationHelperStyles.isAnimating]: isAnimating
        }),
        children: [
          showRecent && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { flexDirection: "column", gap: 0, width: "100%", children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_Text.Text,
              {
                textStyle: "small-emphasized",
                className: styles.sectionHeadline,
                children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  import_react_intl.FormattedMessage,
                  {
                    id: "QZdQoUnAWKJYAyK2",
                    defaultMessage: "Recently used",
                    description: "Headline to group recently used segments presets."
                  }
                )
              }
            ),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_framer_motion.AnimatePresence, { initial: false, children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_ShowMorePanel.ShowMorePanel, { children: [
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ShowMorePanel.ShowMorePanel.PrimaryContent, { children: primaryContent }),
              extendedContent && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ShowMorePanel.ShowMorePanel.ExtendedContent, { children: extendedContent })
            ] }) }, "recent-presets")
          ] }),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_framer_motion.AnimatePresence, { initial: false, children: showPinned && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
            import_Flex.Flex,
            {
              as: import_framer_motion.motion.div,
              flexDirection: "column",
              gap: 4,
              variants: import_SegmentsPreset.expandAndCollapseAnimationVariants,
              initial: "exit",
              animate: "enter",
              exit: "exit",
              onAnimationStart: () => setIsAnimating(true),
              onAnimationComplete: () => setIsAnimating(false),
              children: [
                showRecent && showPinned && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Divider.Divider, {}),
                /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  import_Text.Text,
                  {
                    textStyle: "small-emphasized",
                    className: styles.sectionHeadline,
                    style: { paddingTop: import_spacings.default.Size12 },
                    children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                      import_react_intl.FormattedMessage,
                      {
                        id: "568KZGFcxrq95nEC",
                        defaultMessage: "Pinned in {appName}",
                        description: "Headline to group recently used segments presets.",
                        values: { appName: (0, import_app_environment.getAppName)() }
                      },
                      "pinned-header-text"
                    )
                  },
                  "pinned-header"
                ),
                /* @__PURE__ */ (0, import_jsx_runtime.jsx)("ul", { className: styles.list, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_framer_motion.AnimatePresence, { initial: false, children: data.pinnedPresets.map((preset) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  "li",
                  {
                    className: styles.listItem,
                    "data-dt-name": "Pinned",
                    children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                      import_SegmentsPreset.SegmentsPreset,
                      {
                        uiPreset: preset,
                        id: preset.id,
                        pinLimitReached,
                        isPinned: true,
                        isSelected: (0, import_isSelectedPreset.isSelectedPreset)(segments, preset.segments)
                      }
                    )
                  },
                  `pinned-${preset.id}`
                )) }, "pinned-presets-list") })
              ]
            }
          ) }, "pinned-presets")
        ]
      }
    )
  ] });
};
