// @dynatrace/strato-components/charts/xy-chart/components/layout/XYChartGridLayout.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var XYChartGridLayout_exports = {};
__export(XYChartGridLayout_exports, {
  XYChartGridLayout: () => XYChartGridLayout
});
module.exports = __toCommonJS(XYChartGridLayout_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_ChartGrid_css = require("../../../core/components/chart-grid/ChartGrid.sty.js");
var import_useGridPlotSize = require("../../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_Grid = require("../../../core/components/grid/Grid.js");
var import_detect_scale = require("../../../core/utils/detect-scale.js");
var import_useAxesTicks = require("../../hooks/useAxesTicks.js");
var import_useBarData = require("../../hooks/useBarData.js");
var import_useGetVisibleAxesScales = require("../../hooks/useGetVisibleAxesScales.js");
var import_useShouldShowGrid = require("../../hooks/useShouldShowGrid.js");
const XYChartGridLayout = () => {
  const { width, height } = (0, import_useGridPlotSize.useGridPlotSize)();
  const { shouldShowGrid } = (0, import_useShouldShowGrid.useShouldShowGrid)();
  const visibleAxisScales = (0, import_useGetVisibleAxesScales.useGetVisibleAxesScales)();
  const { topTicks, bottomTicks, leftTicks, rightTicks } = (0, import_useAxesTicks.useAxesTicks)();
  const { hasXGroups, hasYGroups } = (0, import_useBarData.useBarData)();
  if (!shouldShowGrid()) {
    return null;
  }
  if (!visibleAxisScales.hasXAndYScale) {
    return null;
  }
  const xTicks = bottomTicks.numericTicks ?? topTicks.numericTicks;
  const yTicks = leftTicks.numericTicks ?? rightTicks.numericTicks;
  const xCategoryOffset = hasXGroups && (0, import_detect_scale.isBandScale)(visibleAxisScales.xScale) ? visibleAxisScales.xScale.step() / 2 : void 0;
  const yCategoryOffset = hasYGroups && (0, import_detect_scale.isBandScale)(visibleAxisScales.yScale) ? visibleAxisScales.yScale.step() / 2 : void 0;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("svg", { className: import_ChartGrid_css.plotLayout, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Grid.Grid,
    {
      xScale: visibleAxisScales.xScale,
      yScale: visibleAxisScales.yScale,
      height,
      width,
      xTicks,
      yTicks,
      highlightZeroLine: {
        horizontal: false,
        vertical: false
      },
      xCategoryOffset,
      yCategoryOffset
    }
  ) });
};
XYChartGridLayout["displayName"] = "XYChartGridLayout";
