// @dynatrace/strato-components/charts/core/components/overlay-tooltip/components/header/utils/format-date-range.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var format_date_range_exports = {};
__export(format_date_range_exports, {
  formatDateRange: () => formatDateRange
});
module.exports = __toCommonJS(format_date_range_exports);
var import_date_fns = require("date-fns");
var import_units = require("@dynatrace-sdk/units");
var import_format_date_ticks = require("../../../../../utils/format-date-ticks.js");
var import_format_dates = require("../../../../../utils/format-dates.js");
const shouldIncludeSeconds = (date) => date.getSeconds() !== 0;
const createFullDateTimeFormatOptions = (includeSeconds) => ({
  day: "2-digit",
  month: "short",
  year: "numeric",
  hour: "2-digit",
  minute: "2-digit",
  second: includeSeconds ? "numeric" : void 0
});
const formatRange = (from, to, prefix) => {
  const base = `${from} - ${to}`;
  return prefix ? `${prefix}, ${base}` : base;
};
const formatTimeRange = (from, to) => {
  return formatRange((0, import_format_date_ticks.formatDateTicks)(from), (0, import_format_date_ticks.formatDateTicks)(to));
};
const formatMinuteRange = (from, to) => {
  const commonDate = (0, import_units.formatDate)(from, {
    month: "short",
    day: "2-digit",
    year: "numeric"
  });
  const fromTime = (0, import_units.formatDate)(from, { hour: "2-digit", minute: "2-digit" });
  const toTime = (0, import_units.formatDate)(to, { hour: "2-digit", minute: "2-digit" });
  return formatRange(fromTime, toTime, commonDate);
};
const formatHourRange = (from, to) => {
  const commonDay = (0, import_units.formatDate)(from, {
    day: "2-digit",
    month: "short",
    year: "numeric"
  });
  const fromHour = (0, import_units.formatDate)(from, { hour: "2-digit" });
  const toHour = (0, import_units.formatDate)(to, { hour: "2-digit" });
  return formatRange(fromHour, toHour, commonDay);
};
const formatDayRange = (from, to) => {
  const commonMonth = (0, import_units.formatDate)(from, {
    month: "short",
    year: "numeric"
  });
  const fromDay = (0, import_units.formatDate)(from, { day: "2-digit" });
  const toDay = (0, import_units.formatDate)(to, { day: "2-digit" });
  return formatRange(fromDay, toDay, commonMonth);
};
const formatYearRange = (from, to) => {
  const commonYear = (0, import_format_dates.formatYear)(from.getTime());
  const fromWeekday = (0, import_format_dates.formatWeekday)(from.getTime());
  const toWeekday = (0, import_format_dates.formatWeekday)(to.getTime());
  return formatRange(fromWeekday, toWeekday, commonYear);
};
const formatDateRange = (from, to) => {
  if ((0, import_date_fns.isSameMinute)(from, to)) {
    return formatTimeRange(from, to);
  }
  if ((0, import_date_fns.isSameHour)(from, to)) {
    return formatMinuteRange(from, to);
  }
  if ((0, import_date_fns.isSameDay)(from, to)) {
    return formatHourRange(from, to);
  }
  if ((0, import_date_fns.isSameMonth)(from, to)) {
    return formatDayRange(from, to);
  }
  if ((0, import_date_fns.isSameYear)(from, to)) {
    return formatYearRange(from, to);
  }
  const fromFull = (0, import_units.formatDate)(
    from,
    createFullDateTimeFormatOptions(shouldIncludeSeconds(from))
  );
  const toFull = (0, import_units.formatDate)(
    to,
    createFullDateTimeFormatOptions(shouldIncludeSeconds(to))
  );
  return formatRange(fromFull, toFull);
};
