// @dynatrace/strato-components/charts/core/components/overlay-tooltip/components/header/utils/type-guards.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var type_guards_exports = {};
__export(type_guards_exports, {
  isDateRange: () => isDateRange,
  isFromTo: () => isFromTo,
  isHeaderEmpty: () => isHeaderEmpty,
  isNonEmptyRange: () => isNonEmptyRange,
  isNumericRange: () => isNumericRange,
  isSingleValue: () => isSingleValue,
  isStringRange: () => isStringRange
});
module.exports = __toCommonJS(type_guards_exports);
var import_lodash_es = require("lodash");
const isDateRange = (data) => {
  return data instanceof Array && data.length === 2 && data.every((item) => item instanceof Date);
};
const isNumericRange = (data) => {
  return data instanceof Array && data.length === 2 && data.every((item) => typeof item === "number");
};
const isStringRange = (data) => {
  return data instanceof Array && data.length === 2 && data.every((item) => typeof item === "string");
};
const isNonEmptyRange = (data) => {
  return data instanceof Array && data.length === 2 && data.every((item) => !(0, import_lodash_es.isNil)(item));
};
const isFromTo = (data) => {
  return data instanceof Array && data.length === 2 && data.every((item) => !(0, import_lodash_es.isNil)(item));
};
const isSingleValue = (data) => data instanceof Array && data.length === 1 && !(0, import_lodash_es.isNil)(data[0]) || !(data instanceof Array) && !(0, import_lodash_es.isNil)(data);
const isHeaderEmpty = (header1, header2, subheader) => ((0, import_lodash_es.isNil)(header1) || (0, import_lodash_es.isEmpty)(header1)) && ((0, import_lodash_es.isNil)(header2) || (0, import_lodash_es.isEmpty)(header2)) && ((0, import_lodash_es.isNil)(subheader) || (0, import_lodash_es.isEmpty)(subheader));
