// @dynatrace/strato-components/charts/xy-chart/components/overlay/XYChartOverlay.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var XYChartOverlay_exports = {};
__export(XYChartOverlay_exports, {
  XYChartOverlay: () => XYChartOverlay
});
module.exports = __toCommonJS(XYChartOverlay_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_XYChartSynchronizedCrosshair = require("./XYChartSynchronizedCrosshair.js");
var import_useGridPlotSize = require("../../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_InteractionsOverlay = require("../../../core/components/interactions-overlay/InteractionsOverlay.js");
var import_OverlayContext = require("../../../core/components/overlay-tooltip/contexts/OverlayContext.js");
var import_selectors = require("../../store/selectors.js");
var import_XYChartSelectionArea = require("../selection/XYChartSelectionArea.js");
var import_XYChartToolbar = require("../toolbar/XYChartToolbar.js");
var import_XYChartOverlayTooltip = require("../tooltip/XYChartOverlayTooltip.js");
var import_XYChartCrosshair = require("./explore/XYChartCrosshair.js");
var import_XYChartLinearPoint = require("./explore/XYChartLinearPoint.js");
var import_XYChartRectanglePoint = require("./explore/XYChartRectanglePoint.js");
var import_useOverlayEvents = require("./hooks/useOverlayEvents.js");
var import_is_own_event = require("../../../core/components/interactions-overlay/utils/is-own-event.js");
var import_selectors2 = require("../../../core/components/overlay-tooltip/selectors.js");
var import_SynchronizedSelectionArea = require("../../../core/components/shared-chart-interactions/components/SynchronizedSelectionArea.js");
var import_useSynchronizedSelection = require("../../../core/components/shared-chart-interactions/hooks/useSynchronizedSelection.js");
var import_useToolbarHandlerRef = require("../../../core/components/toolbar/hooks/useToolbarHandlerRef.js");
var import_xy_chart_z_index_areas = require("../../constants/xy-chart-z-index-areas.js");
var import_XYChartSelectionAreaRef_context = require("../../context/XYChartSelectionAreaRef.context.js");
var import_useSyncedXScaleByPriority = require("../../hooks/useSyncedXScaleByPriority.js");
var import_useXYChartData = require("../../hooks/useXYChartData.js");
var import_useXYChartId = require("../../hooks/useXYChartId.js");
const XYChartOverlay = () => {
  const setState = (0, import_selectors.useSetState)();
  const handlerRef = (0, import_useToolbarHandlerRef.useToolbarHandlerRef)();
  const overlay = (0, import_OverlayContext.useOverlay)();
  const selectionRef = (0, import_react.useRef)(null);
  const { width, height } = (0, import_useGridPlotSize.useGridPlotSize)();
  (0, import_selectors2.useSelector)(({ inBounds: inBounds2, pinned: pinned2 }) => `${inBounds2}-${pinned2}`);
  const { inBounds, pinned } = (0, import_selectors2.useTooltipSnapshot)();
  const {
    handleModeChange,
    handleHover,
    handleTogglePin,
    handleReset,
    handleLeave,
    handlePanChange,
    handleSelectChange,
    handleSelectEnd,
    handleZoomChange,
    handleSelectStart,
    resetTooltip
  } = (0, import_useOverlayEvents.useOverlayEvents)();
  const mode = (0, import_selectors.useInteractionModeState)();
  const selectState = (0, import_selectors.useSelectState)();
  const synchronizedSelection = (0, import_useSynchronizedSelection.useSynchronizedSelection)();
  const chartId = (0, import_useXYChartId.useXYChartId)();
  const syncedXScale = (0, import_useSyncedXScaleByPriority.useSyncedXScaleByPriority)();
  const { axesMetadata } = (0, import_useXYChartData.useXYChartData)();
  const { zoomOutEnabled } = (0, import_selectors.useZoomState)();
  const isZoomDisabled = !axesMetadata.isZoomEnabled;
  const isPanDisabled = !(axesMetadata.isPanEnabled && zoomOutEnabled);
  (0, import_react.useEffect)(() => {
    if ((0, import_lodash_es.every)([
      selectState,
      !synchronizedSelection.selectedChartId,
      !synchronizedSelection?.disabled
    ])) {
      setState((s) => ({ ...s, select: void 0 }));
    }
  }, [synchronizedSelection.selectedChartId]);
  (0, import_react.useEffect)(() => {
    if (synchronizedSelection.selectedChartId === chartId && !selectState) {
      synchronizedSelection.resetSelectionState?.();
    }
  }, [!!selectState]);
  const showHighlightedData = selectState === void 0 || selectState?.variant === "select";
  const handleFocus = (event) => {
    if (!event || event.cancelable) {
      return;
    }
    if ((0, import_is_own_event.isOwnEvent)(event)) {
      handlerRef.current.handleFocus();
    }
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_InteractionsOverlay.InteractionsOverlay,
    {
      style: {
        zIndex: import_xy_chart_z_index_areas.zIndicesXYChartAreas.overlays,
        userSelect: selectState?.finished === false ? "none" : "auto"
      },
      ref: overlay.overlayRef,
      mode,
      pinned,
      selectionArea: selectState ? [selectState?.start, selectState?.end] : void 0,
      width,
      height,
      onModeChange: handleModeChange,
      onReset: handleReset,
      onLeave: handleLeave,
      onPanChange: handlePanChange,
      onHover: handleHover,
      onFocus: handleFocus,
      onTogglePin: ({ pin, position }) => handleTogglePin(pin, position),
      onSelectStart: handleSelectStart,
      onSelectChange: handleSelectChange,
      onSelectEnd: (event) => handleSelectEnd(event, selectionRef),
      onZoomChange: handleZoomChange,
      isZoomDisabled,
      isZoomXDisabled: isZoomDisabled,
      isPanDisabled,
      children: [
        (inBounds || pinned || selectState) && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartOverlayTooltip.XYChartOverlayTooltip, {}),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartCrosshair.XYChartCrosshair, { chartRef: overlay.overlayRef }),
          showHighlightedData && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartRectanglePoint.XYChartRectanglePoint, {}),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartLinearPoint.XYChartLinearPoint, {})
          ] })
        ] }),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartSynchronizedCrosshair.XYChartSynchronizedCrosshair, {}),
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_XYChartSelectionAreaRef_context.XYChartSelectionAreaRefContext.Provider, { value: selectionRef, children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartToolbar.XYChartToolbar, { onMouseEnter: resetTooltip, handlerRef }),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartSelectionArea.XYChartSelectionArea, {})
        ] }),
        synchronizedSelection.selectedChartId && synchronizedSelection.selectedChartId !== chartId && syncedXScale && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_SynchronizedSelectionArea.SynchronizedSelectionArea, { scale: syncedXScale })
      ]
    }
  );
};
XYChartOverlay["displayName"] = "XYChartOverlay";
