// @dynatrace/strato-components/forms/toggle-button-group/Item.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Item_exports = {};
__export(Item_exports, {
  Item: () => Item
});
module.exports = __toCommonJS(Item_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_useToggleButtonGroupContext = require("./hooks/useToggleButtonGroupContext.js");
var import_Prefix = require("./Prefix.js");
var styles = __toESM(require("./styles.sty.js"));
var import_Suffix = require("./Suffix.js");
var import_VisuallyHidden = require("../../core/components/visually-hidden/VisuallyHidden.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useId = require("../../core/hooks/useId.js");
var import_useSlots = require("../../core/hooks/useSlots.js");
var import_focusRingStatic_css = require("../../core/styles/focusRingStatic.sty.js");
var import_is_node_environment = require("../../core/utils/is-node-environment.js");
var import_merge_props = require("../../core/utils/merge-props.js");
var import_Flex = require("../../layouts/flex/Flex.js");
var import_Tooltip = require("../../overlays/tooltip/Tooltip.js");
var import_Text = require("../../typography/text/Text.js");
var import_useFormFieldMessagesContext = require("../forms-core/hooks/useFormFieldMessagesContext.js");
var import_useFormFieldValidation = require("../forms-core/hooks/useFormFieldValidation.js");
var import_formfield_messages_state_reducer = require("../forms-core/state/formfield-messages-state-reducer.js");
const Item = (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const fallbackId = (0, import_useId.useId)();
    const {
      id = fallbackId,
      disabled: propDisabled,
      readOnly: propReadOnly,
      tooltip,
      children,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      value,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
    const {
      "data-dt-name": dataDtName = "Toggle",
      ...remainingBehavioralTrackingProps
    } = behavioralTrackingProps;
    const {
      name,
      isRequired,
      hasError,
      isDisabled: disabledContext,
      ariaDisabled,
      selectedValue,
      isReadOnly: isReadOnlyContext,
      onChange
    } = (0, import_useToggleButtonGroupContext.useToggleButtonGroupContext)();
    const labelRef = (0, import_react.useRef)(null);
    const inputRef = (0, import_react.useRef)(null);
    const isDisabled = disabledContext || propDisabled;
    const isReadOnly = isReadOnlyContext || propReadOnly;
    const checked = selectedValue === value;
    const color = hasError ? "critical" : "neutral";
    const { dispatch: messagesContextDispatch } = (0, import_useFormFieldMessagesContext.useFormFieldMessagesContext)();
    const {
      onChange: validationOnChange,
      onInvalid: validationOnInvalid,
      onBlur: validationOnBlur,
      ariaProps
    } = (0, import_useFormFieldValidation.useFormFieldValidation)(inputRef, { isRequired });
    (0, import_react.useEffect)(() => {
      messagesContextDispatch?.({
        type: import_formfield_messages_state_reducer.FormFieldMessagesAction.ADDFORMCONTROLID,
        payload: { controlId: id }
      });
      return () => messagesContextDispatch?.({
        type: import_formfield_messages_state_reducer.FormFieldMessagesAction.REMOVEFORMCONTROLID,
        payload: { controlId: id }
      });
    }, [id, messagesContextDispatch]);
    const [slots, freeform] = (0, import_useSlots.useSlots)(children, {
      prefix: import_Prefix.Prefix,
      suffix: import_Suffix.Suffix
    });
    (0, import_react.useImperativeHandle)(forwardedRef, () => {
      return {
        element: labelRef.current,
        inputRef: inputRef.current,
        focus() {
          inputRef.current?.focus();
        },
        validate() {
          inputRef.current?.checkValidity();
        }
      };
    });
    const scrollTabIntoView = (0, import_react.useCallback)(
      (behavior) => {
        if (!checked || !labelRef.current) {
          return;
        }
        const parent = labelRef.current.parentElement;
        if (!parent) {
          return;
        }
        const parentRect = parent.getBoundingClientRect();
        const targetRect = labelRef.current.getBoundingClientRect();
        const currentLeft = parent.scrollLeft;
        const deltaLeft = targetRect.left - parentRect.left;
        const deltaRight = targetRect.right - parentRect.right;
        let nextLeft = currentLeft;
        if (deltaLeft < 0) {
          nextLeft += deltaLeft;
        } else if (deltaRight > 0) {
          nextLeft += deltaRight;
        }
        parent.scrollTo?.({ left: nextLeft, behavior });
      },
      [checked]
    );
    const handleChange = (0, import_react.useCallback)(
      (e) => {
        if (isReadOnly) {
          return;
        }
        if (value !== void 0) {
          onChange?.(value, e);
        }
        validationOnChange?.();
      },
      [isReadOnly, value, onChange, validationOnChange]
    );
    (0, import_react.useEffect)(() => {
      if (checked && labelRef.current) {
        scrollTabIntoView("instant");
      }
    }, []);
    (0, import_react.useEffect)(() => {
      const parent = labelRef.current?.parentElement;
      if (!parent) {
        return;
      }
      let prevWidth = parent.offsetWidth;
      let debounceTimer;
      const resizeObserver = new window.ResizeObserver((entries) => {
        for (const entry of entries) {
          if (entry.target === parent && parent.offsetWidth !== prevWidth) {
            prevWidth = parent.offsetWidth;
            window.clearTimeout(debounceTimer);
            debounceTimer = window.setTimeout(() => {
              scrollTabIntoView("smooth");
            }, 300);
          }
        }
      });
      resizeObserver.observe(parent);
      scrollTabIntoView("smooth");
      return () => {
        resizeObserver.disconnect();
        window.clearTimeout(debounceTimer);
      };
    }, [checked, scrollTabIntoView]);
    if ((0, import_is_node_environment.isDev)() && tooltip === void 0 && freeform.length === 0) {
      console.warn(
        "A text to indicate the value is required. It can be provided by a tooltip."
      );
    }
    const toggleItem = /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      "label",
      {
        "data-testid": dataTestId,
        htmlFor: id,
        ref: labelRef,
        className: (0, import_clsx.default)(
          styles.item,
          {
            [styles.checked]: checked,
            [styles.disabled]: isDisabled || ariaDisabled,
            [styles.readOnly]: isReadOnly
          },
          (0, import_focusRingStatic_css.focusRingStatic)({
            when: "focus-within-visible",
            disabled: isDisabled || ariaDisabled,
            color
          }),
          consumerClassName
        ),
        style: consumerStyle,
        "data-dt-name": dataDtName,
        ...remainingBehavioralTrackingProps,
        ...restProps,
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_VisuallyHidden.VisuallyHidden, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "input",
            {
              id,
              type: "radio",
              ref: inputRef,
              name,
              value,
              disabled: isDisabled,
              "aria-disabled": ariaDisabled,
              readOnly: isReadOnly,
              "aria-label": tooltip,
              checked,
              onChange: handleChange,
              required: isRequired,
              onInvalid: validationOnInvalid,
              ...(0, import_merge_props.mergeProps)(restProps, {
                ...ariaProps,
                onBlur: validationOnBlur
              })
            }
          ) }),
          slots.prefix,
          freeform.length > 0 && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { flexDirection: "column", gap: 0, className: styles.textContainer, children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_Text.Text,
              {
                as: "span",
                textStyle: checked ? "base-emphasized" : "base",
                className: styles.text,
                children: freeform
              }
            ),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_Text.Text,
              {
                as: "span",
                textStyle: "base-emphasized",
                className: (0, import_clsx.default)(styles.hidden, styles.text),
                "aria-hidden": true,
                children: freeform
              }
            )
          ] }),
          slots.suffix
        ]
      }
    );
    return tooltip ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Tooltip.Tooltip, { text: tooltip, children: toggleItem }) : toggleItem;
  }
);
Item.displayName = "ToggleButtonGroup.Item";
