// @dynatrace/strato-components/forms/toggle-button-group/styles.sty.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var styles_css_exports = {};
__export(styles_css_exports, {
  checked: () => checked,
  container: () => container,
  containerWidths: () => containerWidths,
  disabled: () => disabled,
  disabledGroup: () => disabledGroup,
  divider: () => divider,
  error: () => error,
  hidden: () => hidden,
  item: () => item,
  readOnly: () => readOnly,
  scrollButton: () => scrollButton,
  scrollContainer: () => scrollContainer,
  scrollVariants: () => scrollVariants,
  text: () => text,
  textContainer: () => textContainer
});
module.exports = __toCommonJS(styles_css_exports);
var import_AppRoot_css_ts_vanilla = require("../../core/components/app-root/AppRoot.css");
var import_field_css_ts_vanilla = require("../../styles/field.css");
var import_styles_css_ts_vanilla = require("./styles.css");
var checked = "_w5wkeh1-3-6-0";
var container = "_w5wkeh7-3-6-0";
var containerWidths = { full: "_w5wkeh9-3-6-0", content: "_w5wkeha-3-6-0" };
var disabled = "_w5wkeh3-3-6-0";
var disabledGroup = "_w5wkeh5-3-6-0";
var divider = "_w5wkeh8-3-6-0";
var error = "_w5wkeh2-3-6-0";
var hidden = "_w5wkeh6-3-6-0";
var item = "_w5wkehg-3-6-0";
var readOnly = "_w5wkeh4-3-6-0";
var scrollButton = "_w5wkehf-3-6-0";
var scrollContainer = "_w5wkehb-3-6-0";
var scrollVariants = { both: "_w5wkehc-3-6-0", left: "_w5wkehd-3-6-0", right: "_w5wkehe-3-6-0" };
var text = "_w5wkehh-3-6-0";
var textContainer = "_w5wkehi-3-6-0";
