// @dynatrace/strato-components/forms/toggle-button-group/ToggleButtonGroup.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ToggleButtonGroup_exports = {};
__export(ToggleButtonGroup_exports, {
  ToggleButtonGroup: () => ToggleButtonGroup
});
module.exports = __toCommonJS(ToggleButtonGroup_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_ToggleButtonGroupContext = require("./contexts/ToggleButtonGroupContext.js");
var import_Item = require("./Item.js");
var import_Prefix = require("./Prefix.js");
var styles = __toESM(require("./styles.sty.js"));
var import_Suffix = require("./Suffix.js");
var import_Button = require("../../buttons/button/Button.js");
var import_useContainerSize = require("../../core/hooks/useContainerSize.js");
var import_useControllableState = require("../../core/hooks/useControllableState.js");
var import_useId = require("../../core/hooks/useId.js");
var import_useSlots = require("../../core/hooks/useSlots.js");
var import_is_visual_refresh_enabled = require("../../core/utils/is-visual-refresh-enabled.js");
var import_Divider = require("../../layouts/divider/Divider.js");
var import_Flex = require("../../layouts/flex/Flex.js");
var import_useFormFieldMessagesContext = require("../forms-core/hooks/useFormFieldMessagesContext.js");
var import_useSyncedFormControlProps = require("../forms-core/hooks/useSyncedFormControlProps.js");
var import_formfield_messages_state_reducer = require("../forms-core/state/formfield-messages-state-reducer.js");
const COMPONENT_NAME = "ToggleButtonGroup";
const ToggleButtonGroupRoot = (0, import_react.forwardRef)((props, forwardedRef) => {
  const fallbackName = (0, import_useId.useId)();
  const {
    children,
    id: propId,
    disabled: propDisabled,
    required: propRequired,
    readOnly: propReadOnly,
    value,
    defaultValue,
    name = fallbackName,
    width = "content",
    onChange,
    className: consumerClassName,
    style: consumerStyle,
    "aria-disabled": propAriaDisabled,
    "data-dtrum-allow": dataDtrumAllow,
    "data-dtrum-mask": dataDtrumMask,
    ...remainingProps
  } = props;
  const wrapperRef = (0, import_react.useRef)(null);
  const containerRef = (0, import_react.useRef)(null);
  const [isOverflowingLeft, setIsOverflowingLeft] = (0, import_react.useState)(false);
  const [isOverflowingRight, setIsOverflowingRight] = (0, import_react.useState)(false);
  const {
    disabled: formControlDisabled,
    ariaDisabled,
    required: isRequired,
    id: formControlId,
    readOnly: isReadOnly
  } = (0, import_useSyncedFormControlProps.useSyncedFormControlProps)({
    propDisabled,
    propRequired,
    propId,
    propAriaDisabled,
    propReadOnly
  });
  const isDisabled = propDisabled ?? formControlDisabled;
  const isDisabledOrAriaDisabled = isDisabled || ariaDisabled;
  const [selectedValue, setSelectedValue] = (0, import_useControllableState.useControllableState)({
    defaultValue,
    value
  });
  const { dispatch: messagesContextDispatch, state: messageContextState } = (0, import_useFormFieldMessagesContext.useFormFieldMessagesContext)();
  const hasError = messageContextState.showError;
  (0, import_react.useImperativeHandle)(
    forwardedRef,
    () => {
      const newRef = {
        element: wrapperRef.current,
        focus() {
          const checkedInput = wrapperRef.current?.querySelector(
            "input:checked"
          );
          (checkedInput ?? wrapperRef.current?.querySelector("input"))?.focus();
        },
        validate() {
          const inputs = wrapperRef.current?.querySelectorAll("input");
          inputs?.forEach((input) => input?.checkValidity());
          messagesContextDispatch?.({
            type: import_formfield_messages_state_reducer.FormFieldMessagesAction.SETISTOUCHED,
            payload: { isTouched: true }
          });
        }
      };
      if (!wrapperRef.current) {
        return null;
      }
      return new Proxy(wrapperRef.current, {
        get: function(target, property) {
          if (property in newRef) {
            return newRef[property];
          }
          return target[property];
        }
      });
    },
    [messagesContextDispatch]
  );
  const handleChange = (0, import_react.useCallback)(
    (value2, event) => {
      setSelectedValue?.(value2);
      onChange?.(value2, event);
    },
    [onChange, setSelectedValue]
  );
  const customWidth = width !== "content" && width !== "full" ? width : void 0;
  const [slots] = (0, import_useSlots.useSlots)(children, {
    items: { match: import_Item.Item, multiple: true },
    prefix: import_Prefix.Prefix,
    suffix: import_Suffix.Suffix
  });
  const valueProps = (0, import_react.useMemo)(
    () => ({
      name,
      isDisabled,
      ariaDisabled,
      isReadOnly,
      isRequired,
      hasError,
      selectedValue,
      onChange: handleChange
    }),
    [
      name,
      isDisabled,
      ariaDisabled,
      isReadOnly,
      isRequired,
      hasError,
      selectedValue,
      handleChange
    ]
  );
  const scrollButtonVariant = (0, import_is_visual_refresh_enabled.isVisualRefreshEnabled)() ? "default" : "emphasized";
  const color = hasError && !isDisabledOrAriaDisabled ? "critical" : "neutral";
  const { width: containerWidth } = (0, import_useContainerSize.useContainerSize)(containerRef);
  const updateOverflow = (0, import_react.useCallback)(() => {
    if (containerRef.current) {
      const { scrollLeft, scrollWidth, offsetWidth } = containerRef.current;
      const tolerance = 4;
      setIsOverflowingLeft(scrollLeft > tolerance);
      setIsOverflowingRight(scrollLeft + offsetWidth < scrollWidth - tolerance);
    }
  }, []);
  const slideTabs = (modifier) => {
    const container = containerRef.current;
    if (container) {
      container.scrollBy({
        left: modifier * container.offsetWidth / 2,
        behavior: "smooth"
      });
    }
  };
  (0, import_react.useEffect)(() => {
    updateOverflow();
  }, [containerWidth, updateOverflow]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    "div",
    {
      className: (0, import_clsx.default)(
        styles.container,
        {
          [styles.disabledGroup]: isDisabledOrAriaDisabled,
          [styles.error]: hasError,
          [styles.readOnly]: isReadOnly,
          [styles.containerWidths.content]: width === "content",
          [styles.containerWidths.full]: width === "full"
        },
        consumerClassName
      ),
      style: { maxWidth: customWidth, ...consumerStyle },
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "aria-orientation": "horizontal",
      role: "radiogroup",
      "aria-live": "polite",
      "aria-disabled": ariaDisabled,
      id: formControlId,
      ref: wrapperRef,
      ...remainingProps,
      "data-dt-component": COMPONENT_NAME,
      children: [
        isOverflowingLeft && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_Button.Button,
          {
            className: styles.scrollButton,
            style: { left: "4px", position: "absolute", top: "4px" },
            variant: scrollButtonVariant,
            color,
            onClick: () => slideTabs(-1),
            size: "condensed",
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ChevronLeftIcon, {}) })
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_Flex.Flex,
          {
            ref: containerRef,
            gap: 0,
            onScroll: updateOverflow,
            className: (0, import_clsx.default)(styles.scrollContainer, {
              [styles.scrollVariants.left]: isOverflowingLeft && !isOverflowingRight,
              [styles.scrollVariants.right]: !isOverflowingLeft && isOverflowingRight,
              [styles.scrollVariants.both]: isOverflowingLeft && isOverflowingRight
            }),
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ToggleButtonGroupContext.ToggleButtonGroupContext.Provider, { value: valueProps, children: slots.items?.map((slot, index) => {
              const items = slots.items ?? [];
              const value2 = slot.props.value;
              const isSelected = value2 === selectedValue;
              const isLast = index === items.length - 1;
              const nextItem = items[index + 1];
              const nextItemValue = nextItem?.props?.value;
              const nextIsSelected = nextItemValue === selectedValue;
              return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_react.Fragment, { children: [
                slot,
                !isLast && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  import_Divider.Divider,
                  {
                    orientation: "vertical",
                    className: (0, import_clsx.default)(styles.divider, {
                      [styles.hidden]: isSelected || nextIsSelected
                    })
                  }
                )
              ] }, slot.props.value ?? slot.props.id ?? index);
            }) })
          }
        ),
        isOverflowingRight && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_Button.Button,
          {
            className: styles.scrollButton,
            style: { right: "4px", position: "absolute", top: "4px " },
            variant: scrollButtonVariant,
            color,
            onClick: () => slideTabs(1),
            size: "condensed",
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Suffix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ChevronRightIcon, {}) })
          }
        )
      ]
    }
  );
});
const ToggleButtonGroup = Object.assign(ToggleButtonGroupRoot, {
  Item: import_Item.Item,
  Suffix: import_Suffix.Suffix,
  Prefix: import_Prefix.Prefix
});
