// @dynatrace/strato-components/notifications/notification-settings/hooks/toast-alerts.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var toast_alerts_exports = {};
__export(toast_alerts_exports, {
  noWritePermissionToast: () => noWritePermissionToast,
  notificationGenericErrorToast: () => notificationGenericErrorToast,
  notificationLimitReachedToast: () => notificationLimitReachedToast
});
module.exports = __toCommonJS(toast_alerts_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var styles = __toESM(require("../../../buttons/notify-button/NotifyButton.sty.js"));
var import_Text = require("../../../typography/text/Text.js");
var import_Toast = require("../../toast/Toast.js");
var import_CopyPermissionsButton = require("../components/CopyPermissionsButton.js");
var import_messages = require("../messages.js");
function ToastMessage({ children }) {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { className: styles.toast, children });
}
function noWritePermissionToast(intl, notificationEnabled) {
  (0, import_Toast.showToast)({
    title: intl.formatMessage(import_messages.notificationMessages.noPermissionsHeading, {
      notificationEnabled
    }),
    message: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(ToastMessage, { children: [
      intl.formatMessage(import_messages.notificationMessages.copyPermissions),
      "."
    ] }),
    actions: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_CopyPermissionsButton.CopyPermissionsButton, { readOrWrite: "write" }),
    position: "bottom-left",
    type: "critical"
  });
}
function notificationLimitReachedToast(intl, notificationType) {
  (0, import_Toast.showToast)({
    title: intl.formatMessage(import_messages.notificationMessages.limitReachedToastHeading),
    message: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(ToastMessage, { children: intl.formatMessage(import_messages.notificationMessages.limitReached, {
      notificationType
    }) }),
    position: "bottom-left",
    type: "critical"
  });
}
function notificationGenericErrorToast(intl) {
  (0, import_Toast.showToast)({
    title: intl.formatMessage(import_messages.notificationMessages.genericError),
    message: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(ToastMessage, { children: intl.formatMessage(import_messages.notificationMessages.genericErrorMessage) }),
    position: "bottom-left",
    type: "critical"
  });
}
