// @dynatrace/strato-components/notifications/notification-settings/hooks/useEventNotificationClient.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useEventNotificationClient_exports = {};
__export(useEventNotificationClient_exports, {
  useEventNotificationClient: () => useEventNotificationClient
});
module.exports = __toCommonJS(useEventNotificationClient_exports);
var import_client_notification_v2 = require("@dynatrace-sdk/client-notification-v2");
var import_useNotificationClient = require("./useNotificationClient.js");
function useEventNotificationClient(config) {
  const notifications = (0, import_useNotificationClient.useNotificationClient)("event", config, {
    list: (config2, abortSignal) => import_client_notification_v2.eventNotificationsClient.getEventNotifications({
      ...config2,
      abortSignal
    }),
    create: (config2) => import_client_notification_v2.eventNotificationsClient.createEventNotification({
      body: config2
    }),
    destroy: (id) => import_client_notification_v2.eventNotificationsClient.deleteEventNotification({
      id
    }),
    isErrorEnvelopeError: import_client_notification_v2.isErrorEnvelopeError
  });
  return notifications;
}
