// @dynatrace/strato-components/notifications/notification-settings/hooks/useNotificationSettingsMenuItem.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useNotificationSettingsMenuItem_exports = {};
__export(useNotificationSettingsMenuItem_exports, {
  useNotificationSettingsMenuItem: () => useNotificationSettingsMenuItem
});
module.exports = __toCommonJS(useNotificationSettingsMenuItem_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_Menu = require("../../../navigation/menu/Menu.js");
var import_Modal = require("../../../overlays/modal/Modal.js");
var import_EventNotificationClient = require("../components/EventNotificationClient.js");
var import_ResourceNotificationClient = require("../components/ResourceNotificationClient.js");
var import_messages = require("../messages.js");
var import_NotificationSettings = require("../NotificationSettings.js");
function useNotificationSettingsMenuItem(props) {
  const { type, config, content } = props;
  const [modalOpen, setModalOpen] = (0, import_react.useState)(false);
  const NotificationMenuItem = (0, import_react.useMemo)(() => {
    return () => /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Menu.Menu.Item, { onSelect: () => setModalOpen(true), children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Menu.Menu.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.NotificationIcon, {}) }),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_intl.FormattedMessage, { ...import_messages.notificationMessages.menuItemLabel })
    ] });
  }, []);
  const NotificationSettingsModal = (0, import_react.useMemo)(() => {
    let notificationSettings = null;
    if ((0, import_ResourceNotificationClient.isResourceNotificationConfig)(type, config)) {
      notificationSettings = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_NotificationSettings.NotificationSettings, { type: "resource", config, children: content });
    } else if ((0, import_EventNotificationClient.isEventNotificationConfig)(type, config)) {
      notificationSettings = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_NotificationSettings.NotificationSettings, { type: "event", config, children: content });
    }
    return () => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Modal.Modal,
      {
        show: modalOpen,
        size: "small",
        onDismiss: () => setModalOpen(false),
        title: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_react_intl.FormattedMessage,
          {
            ...import_messages.notificationMessages.notificationSettingsModalTitle
          }
        ),
        children: notificationSettings
      }
    );
  }, [config, content, modalOpen, type]);
  return [NotificationMenuItem, NotificationSettingsModal];
}
