// @dynatrace/strato-components/notifications/notification-settings/hooks/useNotificationClient.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useNotificationClient_exports = {};
__export(useNotificationClient_exports, {
  useNotificationClient: () => useNotificationClient
});
module.exports = __toCommonJS(useNotificationClient_exports);
var import_react_query = require("@tanstack/react-query");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_notification_query_client = require("./notification-query-client.js");
var import_toast_alerts = require("./toast-alerts.js");
var import_NotifyButtonContext = require("../../../buttons/notify-button/contexts/NotifyButtonContext.js");
function useNotificationClient(notificationType, config, connectors) {
  const intl = (0, import_react_intl.useIntl)();
  const notificationQueryClient = (0, import_react.useContext)(import_notification_query_client.NotificationQueryClientContext);
  const { closeOverlay } = (0, import_react.useContext)(import_NotifyButtonContext.NotifyButtonContext);
  const {
    data: notificationData,
    error: notificationListError,
    isLoading: notificationIsLoading
  } = (0, import_react_query.useQuery)(
    {
      initialData: null,
      queryKey: [config, notificationType],
      queryFn: async ({ signal }) => {
        const data = await connectors.list(config, signal);
        const firstResult = data.results.at(0);
        if (firstResult) {
          return firstResult;
        }
        return null;
      }
    },
    notificationQueryClient
  );
  const {
    mutate: createNotification,
    error: notificationCreateError,
    isPending: isCreatePending
  } = (0, import_react_query.useMutation)(
    {
      mutationKey: [config, notificationType, "create"],
      mutationFn: async () => {
        return await connectors.create(config);
      },
      throwOnError: false,
      // Handle various error scenarios to prompt user information.
      onError: (error) => {
        if (error.name === "403") {
          (0, import_toast_alerts.noWritePermissionToast)(intl, notificationData !== null);
          closeOverlay?.();
        } else if (error.name === "400" && connectors.isErrorEnvelopeError(error) && error.body.error.details?.errorCode?.includes(
          "NotificationLimitReached"
        )) {
          (0, import_toast_alerts.notificationLimitReachedToast)(intl, notificationType);
          closeOverlay?.();
        } else {
          (0, import_toast_alerts.notificationGenericErrorToast)(intl);
        }
      },
      // Update the query cache to avoid the necessary refetch. The create service
      // already responds with the created notification, so we can use that one in the
      // state.
      onSuccess: (data) => {
        notificationQueryClient.setQueryData([config, notificationType], data);
      }
    },
    notificationQueryClient
  );
  const {
    mutate: deleteNotification,
    error: notificationDeleteError,
    isPending: isDeletePending
  } = (0, import_react_query.useMutation)(
    {
      mutationKey: [config, notificationType, "delete"],
      mutationFn: async () => {
        if (notificationData?.id) {
          return connectors.destroy(notificationData.id);
        }
      },
      throwOnError: false,
      onSuccess: (data) => {
        notificationQueryClient.setQueryData([config, notificationType], null);
      },
      onError: (error) => {
        if (error.name === "403") {
          (0, import_toast_alerts.noWritePermissionToast)(intl, notificationData !== null);
          closeOverlay?.();
        }
      }
    },
    notificationQueryClient
  );
  const toggleNotifications = (0, import_react.useCallback)(async () => {
    closeOverlay?.();
    if (notificationData?.id) {
      deleteNotification();
    } else {
      createNotification();
    }
  }, [
    closeOverlay,
    createNotification,
    deleteNotification,
    notificationData?.id
  ]);
  return {
    toggleNotifications,
    enabled: notificationData !== null,
    loading: notificationIsLoading || isCreatePending || isDeletePending,
    hasNotFoundError: notificationListError?.name === "404",
    hasReadPermission: notificationListError?.name !== "403",
    hasWritePermission: notificationCreateError?.name !== "403" && notificationDeleteError?.name !== "403"
  };
}
