// @dynatrace/strato-components/charts/core/components/sparkline-display/constants.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var constants_exports = {};
__export(constants_exports, {
  AXIS_HEIGHT: () => AXIS_HEIGHT,
  CONTEXT_VALUES_FONT: () => CONTEXT_VALUES_FONT,
  CONTEXT_VALUES_Y_OFFSET: () => CONTEXT_VALUES_Y_OFFSET,
  CONTEXT_VALUE_INDICATOR_X_OFFSET: () => CONTEXT_VALUE_INDICATOR_X_OFFSET,
  CONTEXT_VALUE_LABELS_GAP: () => CONTEXT_VALUE_LABELS_GAP,
  CONTEXT_VALUE_LARGE_LINE_HEIGHT: () => CONTEXT_VALUE_LARGE_LINE_HEIGHT,
  CONTEXT_VALUE_LINE_HEIGHT: () => CONTEXT_VALUE_LINE_HEIGHT,
  CONTEXT_VALUE_MAX_WIDTH: () => CONTEXT_VALUE_MAX_WIDTH,
  CONTEXT_VALUE_MIN_LABELS_DISTANCE_LARGE: () => CONTEXT_VALUE_MIN_LABELS_DISTANCE_LARGE,
  CONTEXT_VALUE_MIN_LABELS_DISTANCE_SMALL: () => CONTEXT_VALUE_MIN_LABELS_DISTANCE_SMALL,
  CONTEXT_VALUE_SIGN_COLUMN_HEIGHT_THRESHOLD: () => CONTEXT_VALUE_SIGN_COLUMN_HEIGHT_THRESHOLD,
  CONTEXT_VALUE_SIGN_COLUMN_LARGE: () => CONTEXT_VALUE_SIGN_COLUMN_LARGE,
  CONTEXT_VALUE_SIGN_COLUMN_SMALL: () => CONTEXT_VALUE_SIGN_COLUMN_SMALL,
  CONTEXT_VALUE_SIGN_NUDGE_RIGHT: () => CONTEXT_VALUE_SIGN_NUDGE_RIGHT,
  CONTEXT_VALUE_STROKE_COLOR: () => CONTEXT_VALUE_STROKE_COLOR,
  CONTEXT_VALUE_STROKE_DASHARRAY: () => CONTEXT_VALUE_STROKE_DASHARRAY,
  CONTEXT_VALUE_STROKE_LINECAP: () => CONTEXT_VALUE_STROKE_LINECAP,
  CONTEXT_VALUE_TEXT_OFFSET_MAX: () => CONTEXT_VALUE_TEXT_OFFSET_MAX,
  CONTEXT_VALUE_TEXT_OFFSET_MIN_LARGE: () => CONTEXT_VALUE_TEXT_OFFSET_MIN_LARGE,
  CONTEXT_VALUE_TEXT_OFFSET_MIN_SMALL: () => CONTEXT_VALUE_TEXT_OFFSET_MIN_SMALL,
  CONTEXT_VALUE_TRAILING_PADDING: () => CONTEXT_VALUE_TRAILING_PADDING,
  DEFAULT_SCALE_FALLBACK_COLOR: () => DEFAULT_SCALE_FALLBACK_COLOR,
  DEFAULT_SHOW_CONTEXT_VALUES: () => DEFAULT_SHOW_CONTEXT_VALUES,
  FONT_SIZE_BREAKING_POINT: () => FONT_SIZE_BREAKING_POINT,
  NEGATIVE_SIGN: () => NEGATIVE_SIGN,
  SPARKLINE_ARIA_LABEL: () => SPARKLINE_ARIA_LABEL,
  TIMESTAMP_BAR_WIDTH: () => TIMESTAMP_BAR_WIDTH,
  TIMESTAMP_BAR_X_DEVIATION: () => TIMESTAMP_BAR_X_DEVIATION,
  defaultXAxisConfig: () => defaultXAxisConfig,
  defaultYAxisConfig: () => defaultYAxisConfig
});
module.exports = __toCommonJS(constants_exports);
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
var import_typography = __toESM(require("@dynatrace/strato-design-tokens/typography"));
const defaultXAxisConfig = {
  min: "data-min",
  max: "data-max"
};
const defaultYAxisConfig = {
  min: "data-min",
  max: "data-max",
  scale: "linear"
};
const DEFAULT_SCALE_FALLBACK_COLOR = import_colors.default.Text.Neutral.Default;
const DEFAULT_SHOW_CONTEXT_VALUES = false;
const TIMESTAMP_BAR_WIDTH = 5;
const TIMESTAMP_BAR_X_DEVIATION = 2;
const CONTEXT_VALUE_STROKE_DASHARRAY = "2,2";
const CONTEXT_VALUE_STROKE_LINECAP = "round";
const CONTEXT_VALUE_STROKE_COLOR = import_colors.default.Background.Field.Neutral.DefaultActive;
const CONTEXT_VALUES_Y_OFFSET = 0;
const AXIS_HEIGHT = 25;
const CONTEXT_VALUE_LINE_HEIGHT = 12;
const CONTEXT_VALUE_LARGE_LINE_HEIGHT = 16;
const CONTEXT_VALUE_TEXT_OFFSET_MAX = 4;
const CONTEXT_VALUE_TEXT_OFFSET_MIN_SMALL = 8;
const CONTEXT_VALUE_TEXT_OFFSET_MIN_LARGE = 12;
const CONTEXT_VALUE_LABELS_GAP = 8;
const CONTEXT_VALUE_MIN_LABELS_DISTANCE_SMALL = CONTEXT_VALUE_TEXT_OFFSET_MIN_SMALL + CONTEXT_VALUE_LINE_HEIGHT + CONTEXT_VALUE_LABELS_GAP - CONTEXT_VALUE_TEXT_OFFSET_MAX;
const CONTEXT_VALUE_MIN_LABELS_DISTANCE_LARGE = CONTEXT_VALUE_TEXT_OFFSET_MIN_LARGE + CONTEXT_VALUE_LARGE_LINE_HEIGHT + CONTEXT_VALUE_LABELS_GAP - CONTEXT_VALUE_TEXT_OFFSET_MAX;
const CONTEXT_VALUES_FONT = import_typography.default.Text.Small.Emphasized;
const CONTEXT_VALUE_MAX_WIDTH = 300;
const SPARKLINE_ARIA_LABEL = "Sparkline";
const CONTEXT_VALUE_INDICATOR_X_OFFSET = 8;
const CONTEXT_VALUE_SIGN_COLUMN_SMALL = 8;
const CONTEXT_VALUE_SIGN_COLUMN_LARGE = 10;
const NEGATIVE_SIGN = "-";
const CONTEXT_VALUE_SIGN_NUDGE_RIGHT = 5;
const CONTEXT_VALUE_TRAILING_PADDING = 3;
const CONTEXT_VALUE_SIGN_COLUMN_HEIGHT_THRESHOLD = 24;
const FONT_SIZE_BREAKING_POINT = 32;
