// @dynatrace/strato-components/charts/core/components/sparkline-display/SparklineRenderer.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SparklineRenderer_exports = {};
__export(SparklineRenderer_exports, {
  SparklineRenderer: () => SparklineRenderer
});
module.exports = __toCommonJS(SparklineRenderer_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_accessors = require("./utils/accessors.js");
var import_should_display_scatter_plot = require("./utils/should-display-scatter-plot.js");
var import_AreaCanvas = require("../canvas/components/AreaCanvas.js");
var import_BarCanvas = require("../canvas/components/BarCanvas.js");
var import_LineCanvas = require("../canvas/components/LineCanvas.js");
const SparklineRenderer = ({
  data,
  width,
  height,
  xScale,
  yScale,
  fillColor,
  strokeColor,
  variant,
  strokeWidth,
  curve,
  radius,
  gapPolicy
}) => {
  const areaColor = typeof fillColor === "string" ? () => fillColor : () => ({ context }) => fillColor({ context });
  return variant === "area" ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_AreaCanvas.AreaCanvas,
    {
      data,
      canvasWidth: width,
      canvasHeight: height,
      x: (0, import_accessors.xAxisAccessor)(xScale),
      y1: (0, import_accessors.y1Accessor)(yScale),
      y0: (0, import_accessors.y0Accessor)(yScale),
      defined: import_accessors.definedAccessor,
      color: areaColor,
      displayScatterPlot: (0, import_should_display_scatter_plot.shouldDisplayScatterPlot)(gapPolicy),
      lineColor: () => strokeColor,
      curve
    }
  ) : variant === "line" ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_LineCanvas.LineCanvas,
    {
      data,
      canvasWidth: width,
      canvasHeight: height,
      x: (0, import_accessors.xAxisAccessor)(xScale),
      y: (0, import_accessors.yAccessor)(yScale),
      defined: import_accessors.definedAccessor,
      color: () => strokeColor,
      displayScatterPlot: (0, import_should_display_scatter_plot.shouldDisplayScatterPlot)(gapPolicy),
      strokeWidth,
      curve,
      radius
    }
  ) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_BarCanvas.BarCanvas,
    {
      data,
      canvasWidth: width,
      canvasHeight: height,
      x: (0, import_accessors.barXAxisAccessor)(xScale),
      y: (0, import_accessors.barYAxisAccessor)(yScale),
      width: (point) => (0, import_accessors.barWidthAccessor)(point, xScale),
      height: (point) => (0, import_accessors.barHeightAccessor)(point, yScale),
      strokeWidth: 1,
      color: () => strokeColor,
      opacity: () => 1
    }
  );
};
