// @dynatrace/strato-components/charts/core/components/sparkline-display/SparklineContainer.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SparklineContainer_exports = {};
__export(SparklineContainer_exports, {
  SparklineContainer: () => SparklineContainer
});
module.exports = __toCommonJS(SparklineContainer_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_shape = require("@visx/shape");
var import_ContextValueIndicator = require("./components/ContextValueIndicator.js");
var import_constants = require("./constants.js");
var import_SparklineRenderer = require("./SparklineRenderer.js");
var import_sparkline_container_height = require("./types/sparkline-container-height.js");
var import_arc_radius = require("./utils/arc-radius.js");
var import_get_adjusted_context_value_y_position = require("./utils/get-adjusted-context-value-y-position.js");
var import_get_context_value_layout = require("./utils/get-context-value-layout.js");
var import_get_formatted_context_values = require("./utils/get-formatted-context-values.js");
var import_get_show_context_values = require("./utils/get-show-context-values.js");
var import_get_sparkline_height = require("./utils/get-sparkline-height.js");
var import_get_sparkline_x_scale = require("./utils/get-sparkline-x-scale.js");
var import_get_sparkline_y_scale = require("./utils/get-sparkline-y-scale.js");
var import_get_x_axis_config = require("./utils/get-x-axis-config.js");
var import_map_container_height_to_variant = require("./utils/map-container-height-to-variant.js");
var import_sanitize_sparkline_data = require("./utils/sanitize-sparkline-data.js");
var import_stroke_width = require("./utils/stroke-width.js");
var import_threshold_indicators = require("./utils/threshold-indicators.js");
var import_useCurrentTheme = require("../../../../core/hooks/useCurrentTheme.js");
var import_useObserveContainerSize = require("../../hooks/useObserveContainerSize.js");
var import_formatter_by_ticks = require("../../utils/formatter-by-ticks.js");
var import_getValidColor = require("../../utils/getValidColor.js");
var import_ticks = require("../../utils/ticks.js");
var import_TimeBottomAxis = require("../axes/components/TimeBottomAxis.js");
var import_Canvas = require("../canvas/Canvas.js");
var import_CanvasContext = require("../canvas/CanvasContext.js");
var import_create_canvas_gradient = require("../canvas/utils/create-canvas-gradient.js");
var import_GridLine = require("../grid/GridLine.js");
var import_ThresholdIndicator = require("../threshold-bar/indicators/ThresholdIndicator.js");
function getYPosition(height, adjustedY) {
  return Math.max(0, Math.min(height, adjustedY));
}
const SparklineContainer = (props) => {
  const {
    data,
    color,
    variant = "line",
    showTicks = false,
    thresholds = [],
    gapPolicy = "connect",
    showContextValues = import_constants.DEFAULT_SHOW_CONTEXT_VALUES,
    curve,
    ticksColor,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    className,
    style: consumerStyle,
    labelsAlignment = "left",
    labelsFixedWidth
  } = props;
  const [containerDimension, sizeObserverRef] = (0, import_useObserveContainerSize.useObserveContainerSize)();
  const theme = (0, import_useCurrentTheme.useCurrentTheme)();
  const shouldShowTicks = showTicks && containerDimension.height >= 40;
  const { width: containerWidth } = containerDimension;
  const xAxis = (0, import_get_x_axis_config.getXAxisConfig)(props.xAxis);
  const yAxis = {
    ...import_constants.defaultYAxisConfig,
    ...props.yAxis
  };
  const { sanitizedSeries, domainValues } = (0, import_sanitize_sparkline_data.sanitizeSparklineData)(
    data,
    gapPolicy,
    xAxis
  );
  const { showContextMin, showContextMax } = showContextValues ? (0, import_get_show_context_values.getShowContextValues)(yAxis, domainValues) : { showContextMin: false, showContextMax: false };
  const height = (0, import_get_sparkline_height.getSparklineHeight)(containerDimension.height, shouldShowTicks);
  const isBarVariant = variant === "bar";
  const yScale = (0, import_get_sparkline_y_scale.getSparklineYScale)(domainValues, height, yAxis, isBarVariant);
  const { formattedContextMaxValue, formattedContextMinValue } = (0, import_get_formatted_context_values.getFormattedContextValues)({
    unit: data?.unit,
    formatter: props.yAxis?.formatter,
    domainValues
  });
  const { chartLeftOffset, chartWidth } = (0, import_get_context_value_layout.getContextValueLayout)({
    containerHeight: containerDimension.height,
    containerWidth,
    showContextValues,
    formattedContextMaxValue,
    formattedContextMinValue,
    labelsAlignment,
    labelsFixedWidth
  });
  const strokeWidth = (0, import_stroke_width.calculateStrokeWidth)(height, chartWidth);
  const arcRadius = (0, import_arc_radius.calculateArcRadius)(height, chartWidth);
  const xScale = (0, import_get_sparkline_x_scale.getSparklineXScale)(
    sanitizedSeries,
    chartWidth,
    strokeWidth / 2,
    isBarVariant
  );
  const horizontalNumTicks = (0, import_ticks.getHorizontalNumTicks)(chartWidth);
  const { ticks, formattedTicks } = (0, import_formatter_by_ticks.getFormattedTicks)(
    horizontalNumTicks,
    xScale,
    void 0,
    void 0
  );
  const thresholdIndicators = (0, import_threshold_indicators.mapToThresholdsIndicators)(
    thresholds,
    chartWidth,
    height,
    xScale,
    yScale
  );
  const validatedColor = (0, import_getValidColor.getValidColor)(color, import_constants.DEFAULT_SCALE_FALLBACK_COLOR);
  const [min, max] = yScale.domain();
  const displayZeroLine = min <= 0 && max >= 0;
  const zeroLineYPos = yScale(0) ?? -1;
  const rawMinY = yScale(domainValues.min);
  const rawMaxY = yScale(domainValues.max);
  const containerHeight = (0, import_map_container_height_to_variant.mapContainerHeightToVariant)(
    containerDimension.height
  );
  const fontSize = containerDimension.height >= import_constants.FONT_SIZE_BREAKING_POINT ? import_sparkline_container_height.SparklineContainerHeightFontSize.GREATER_OR_EQUAL_32 : import_sparkline_container_height.SparklineContainerHeightFontSize.SMALLER_THAN_32;
  const minLabelsDistance = fontSize === import_sparkline_container_height.SparklineContainerHeightFontSize.GREATER_OR_EQUAL_32 ? import_constants.CONTEXT_VALUE_MIN_LABELS_DISTANCE_LARGE : import_constants.CONTEXT_VALUE_MIN_LABELS_DISTANCE_SMALL;
  const maxTextOffset = import_constants.CONTEXT_VALUE_TEXT_OFFSET_MAX;
  const minTextOffset = fontSize === import_sparkline_container_height.SparklineContainerHeightFontSize.SMALLER_THAN_32 ? import_constants.CONTEXT_VALUE_TEXT_OFFSET_MIN_SMALL : import_constants.CONTEXT_VALUE_TEXT_OFFSET_MIN_LARGE;
  const { adjustedMinLabelY, adjustedMaxLabelY } = (0, import_get_adjusted_context_value_y_position.getContextValueYPositions)({
    rawMinY,
    rawMaxY,
    showContextMin,
    showContextMax,
    minLabelsDistance,
    minTextOffset,
    maxTextOffset,
    chartHeight: height
  });
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    "div",
    {
      style: { height: "100%" },
      className,
      ref: sizeObserverRef,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "aria-label": import_constants.SPARKLINE_ARIA_LABEL,
      role: "img",
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
          "div",
          {
            style: {
              transform: `translate(${chartLeftOffset}px)`,
              display: "flex",
              flexDirection: "column"
            },
            children: [
              /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
                "svg",
                {
                  width: chartWidth,
                  height,
                  style: { position: "absolute" },
                  children: [
                    displayZeroLine && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                      import_shape.Line,
                      {
                        from: { x: 0, y: zeroLineYPos },
                        to: { x: chartWidth, y: zeroLineYPos },
                        stroke: import_GridLine.DEFAULT_STROKE_COLOR,
                        strokeWidth: import_GridLine.DEFAULT_STROKE_WIDTH,
                        "data-testid": "sparkline-zero-baseline"
                      }
                    ),
                    thresholdIndicators.length > 0 && thresholdIndicators.map(
                      ({ id, indicatorData, color: color2, strokeOnly }) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                        import_ThresholdIndicator.ThresholdIndicator,
                        {
                          data: indicatorData,
                          color: color2,
                          strokeOnly,
                          isActive: false
                        },
                        `threshold-indicator-${id}`
                      )
                    )
                  ]
                }
              ),
              /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { style: { position: "absolute", ...consumerStyle }, children: [
                /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  import_CanvasContext.CanvasContext.Provider,
                  {
                    value: {
                      shapeRefs: /* @__PURE__ */ new Map()
                    },
                    children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Canvas.Canvas, { width: containerWidth, height, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                      import_SparklineRenderer.SparklineRenderer,
                      {
                        data: sanitizedSeries.datapoints,
                        width: chartWidth,
                        height,
                        xScale,
                        yScale,
                        strokeColor: validatedColor,
                        fillColor: (0, import_create_canvas_gradient.createCanvasGradient)({
                          height,
                          color: validatedColor,
                          domain: yScale.domain(),
                          theme
                        }),
                        variant,
                        strokeWidth,
                        radius: arcRadius,
                        curve,
                        gapPolicy
                      }
                    ) })
                  }
                ),
                shouldShowTicks && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  import_TimeBottomAxis.TimeBottomAxis,
                  {
                    ticks,
                    tickColor: ticksColor,
                    formattedTicks,
                    scale: xScale,
                    containerHeight: import_constants.AXIS_HEIGHT,
                    containerWidth: chartWidth - 16
                  }
                )
              ] })
            ]
          }
        ),
        showContextMin && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_ContextValueIndicator.ContextValueIndicator,
          {
            formattedValue: formattedContextMinValue,
            dottedLineWidth: chartWidth,
            labelsFixedWidth,
            yPosition: getYPosition(height, rawMinY),
            labelYPosition: adjustedMinLabelY,
            containerHeight,
            fontSize,
            labelsAlignment,
            chartLeftOffset,
            color: ticksColor,
            sparklineHeight: height
          }
        ),
        showContextMax && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_ContextValueIndicator.ContextValueIndicator,
          {
            formattedValue: formattedContextMaxValue,
            dottedLineWidth: chartWidth,
            yPosition: getYPosition(height, rawMaxY),
            labelYPosition: adjustedMaxLabelY,
            containerHeight,
            fontSize,
            labelsAlignment,
            chartLeftOffset,
            labelsFixedWidth,
            color: ticksColor,
            sparklineHeight: height,
            isMax: true
          }
        )
      ]
    }
  );
};
SparklineContainer["displayName"] = "SparklineContainer";
