// @dynatrace/strato-components/charts/core/components/sparkline-display/SparklineDisplay.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SparklineDisplay_exports = {};
__export(SparklineDisplay_exports, {
  SparklineDisplay: () => SparklineDisplay
});
module.exports = __toCommonJS(SparklineDisplay_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_SparklineContainer = require("./SparklineContainer.js");
var import_get_wrapper_style = require("./utils/get-wrapper-style.js");
var import_useBehavioralTrackingProps = require("../../../../core/hooks/useBehavioralTrackingProps.js");
var import_defaultChartValues_css = require("../../../../core/styles/defaultChartValues.sty.js");
var import_ChartStateHandler = require("../chart-states/ChartStateHandler.js");
var import_get_chart_states_template = require("../chart-states/utils/get-chart-states-template.js");
const COMPONENT_NAME = "Sparkline";
const SparklineDisplay = (props) => {
  const {
    data,
    loading = false,
    children,
    width,
    height,
    style,
    className,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "data-testid": dataTestId
  } = props;
  const [behavioralTrackingProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(props);
  const series = Array.isArray(data) ? data[0] : data;
  const { emptyState, errorState } = (0, import_get_chart_states_template.getChartStatesTemplates)(children);
  const isEmpty = (0, import_lodash_es.isNil)(series) || series.datapoints?.length === 0;
  const wrapperStyle = (0, import_get_wrapper_style.getWrapperStyle)(width, height, style);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ChartStateHandler.ChartStateHandler,
    {
      ...behavioralTrackingProps,
      "data-testid": dataTestId ?? "sparkline-state-handler",
      empty: isEmpty,
      height: "100%",
      emptyState,
      errorState,
      isLoading: loading,
      className: className ?? import_defaultChartValues_css.defaultChartSize,
      style: wrapperStyle,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "data-dt-component": COMPONENT_NAME,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_SparklineContainer.SparklineContainer, { ...props, data: series })
    }
  );
};
SparklineDisplay["displayName"] = "SparklineDisplay";
