// @dynatrace/strato-components/charts/tree-map/components/TreeMapLegend/TreeMapRangeLegend.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TreeMapRangeLegend_exports = {};
__export(TreeMapRangeLegend_exports, {
  TreeMapRangeLegend: () => TreeMapRangeLegend
});
module.exports = __toCommonJS(TreeMapRangeLegend_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_useChartLayoutOrientation = require("../../../core/components/chart-layout/hooks/useChartLayoutOrientation.js");
var import_orientation = require("../../../core/components/chart-layout/utils/orientation.js");
var import_RangeLegend = require("../../../core/components/legend/range/RangeLegend.js");
var import_compute_range_legend_formatter = require("../../../core/utils/compute-range-legend-formatter.js");
var import_FormatterContext = require("../../context/FormatterContext.js");
var import_TreeMapInternalDataContext = require("../../context/TreeMapInternalDataContext.js");
var import_selectors = require("../../store/selectors.js");
const TreeMapRangeLegend = () => {
  const legendColorMap = (0, import_selectors.useColorMap)();
  const { orientation: layoutOrientation } = (0, import_useChartLayoutOrientation.useChartLayoutOrientation)();
  const legendOrientation = (0, import_orientation.getLegendOrientation)(layoutOrientation);
  const { unit } = (0, import_TreeMapInternalDataContext.useTreeMapInternalData)();
  const formatter = (0, import_FormatterContext.useFormatter)();
  const computedFormatter = (0, import_compute_range_legend_formatter.computeRangeLegendFormatter)(formatter, unit);
  const legendItemKeys = [...legendColorMap.keys()];
  const items = legendItemKeys.map((key) => ({
    from: key.from,
    to: key.to,
    color: legendColorMap.get(key)?.color?.Default
  }));
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_RangeLegend.RangeLegend,
    {
      items,
      orientation: legendOrientation,
      formatter: computedFormatter
    }
  );
};
