// @dynatrace/strato-components/esm/charts/xy-chart/store/reducer-explorer.js
function reduceExploreReset(prev) {
  return {
    ...prev,
    highlightedSeriesId: void 0,
    select: void 0
  };
}
function reduceExplorePin(highlightedSeriesId) {
  return (prev) => ({
    ...prev,
    highlightedSeriesId
  });
}
function reduceExploreUnPin(prev) {
  return {
    ...prev,
    select: void 0,
    highlightedSeriesId: void 0
  };
}
function reduceExploreSelect(select) {
  return (prev) => ({
    ...prev,
    select: {
      ...prev.select,
      ...select
    },
    highlightedSeriesId: void 0
  });
}
function reduceExploreCancelSelect(prev) {
  return {
    ...prev,
    select: void 0
  };
}
function reduceExploreCancelHighlight(prev) {
  return {
    ...prev,
    highlightedSeriesId: void 0
  };
}
function reduceExploreCancelAll(prev) {
  return {
    ...prev,
    select: void 0,
    highlightedSeriesId: void 0
  };
}
export {
  reduceExploreCancelAll,
  reduceExploreCancelHighlight,
  reduceExploreCancelSelect,
  reduceExplorePin,
  reduceExploreReset,
  reduceExploreSelect,
  reduceExploreUnPin
};
