// @dynatrace/strato-components/esm/charts/xy-chart/store/selectors.js
import { isEqual } from "lodash-es";
import { useContext } from "react";
import { useStoreSelector } from "../../../core/store/selector.js";
import { XYChartStoreContext } from "../context/XYChartStore.context.js";
const useSelector = (selector, compareFn) => useStoreSelector(XYChartStoreContext, selector, compareFn);
const useSetState = () => {
  const { setState } = useContext(XYChartStoreContext);
  return setState;
};
const useXYChartHighlightedSeriesIdsState = () => useSelector((state) => state.highlightedSeriesId, isEqual);
const useInteractionModeState = () => useSelector((state) => state.interactionMode);
const useZoomState = () => useSelector((state) => state.zoom, isEqual);
const useCurrentDomainState = () => useSelector((state) => state.currentDomain, isEqual);
const useSelectState = () => useSelector((state) => state.select);
const useHoveredThresholdState = () => useSelector((state) => state.thresholds.hoveredThresholdId);
const useAnnotationsState = () => useSelector((state) => state.annotations);
const areSetsEqual = (a, b) => {
  if (a === b) {
    return true;
  }
  if (!a || !b) {
    return false;
  }
  if (a.size !== b.size) {
    return false;
  }
  for (const value of a) {
    if (!b.has(value)) {
      return false;
    }
  }
  return true;
};
const useFilteredSeriesIdsState = () => useSelector((state) => state.filteredSeriesIds, areSetsEqual);
export {
  useAnnotationsState,
  useCurrentDomainState,
  useFilteredSeriesIdsState,
  useHoveredThresholdState,
  useInteractionModeState,
  useSelectState,
  useSelector,
  useSetState,
  useXYChartHighlightedSeriesIdsState,
  useZoomState
};
