// @dynatrace/strato-components/charts/gauge/components/GaugeChartLayout/GaugeChartLayout.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var GaugeChartLayout_exports = {};
__export(GaugeChartLayout_exports, {
  GaugeChartLayout: () => GaugeChartLayout
});
module.exports = __toCommonJS(GaugeChartLayout_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_GaugeChartLayout_css = require("./GaugeChartLayout.sty.js");
var import_access_value = require("../../../../core/utils/accessors/access-value.js");
var import_constants = require("../../../core/constants.js");
var import_useGaugeChartColorContext = require("../../hooks/useGaugeChartColorContext.js");
var import_useGaugeChartLoadingContext = require("../../hooks/useGaugeChartLoadingContext.js");
var import_useGaugeChartRangeMaxContext = require("../../hooks/useGaugeChartRangeMaxContext.js");
var import_useGaugeChartRangeMinContext = require("../../hooks/useGaugeChartRangeMinContext.js");
var import_useGaugeChartThresholdContext = require("../../hooks/useGaugeChartThresholdContext.js");
var import_GaugeBaseArc = require("../GaugeBaseArc.js");
var import_GaugeLabels = require("../GaugeLabels.js");
var import_GaugeLayoutOverlay = require("../GaugeLayoutOverlay/GaugeLayoutOverlay.js");
var import_GaugeRootSVG = require("../GaugeRootSVG.js");
var import_GaugeThresholdIndicators = require("../GaugeThresholdIndicators.js");
var import_GaugeValueArc = require("../GaugeValueArc.js");
const GaugeChartLayout = (0, import_react.forwardRef)(({ value, prefixIcon, intents, toolbar, valueAccessor }, forwardedRef) => {
  const thresholds = (0, import_useGaugeChartThresholdContext.useGaugeChartThresholdContext)();
  const min = (0, import_useGaugeChartRangeMinContext.useGaugeChartRangeMinContext)();
  const max = (0, import_useGaugeChartRangeMaxContext.useGaugeChartRangeMaxContext)();
  const valueArcColor = (0, import_useGaugeChartColorContext.useGaugeChartColorContext)();
  const showLoading = (0, import_useGaugeChartLoadingContext.useGaugeChartLoadingContext)();
  const displayValue = typeof value === "number" ? value : (0, import_access_value.accessValue)(value, valueAccessor ?? import_constants.DEFAULT_VALUE_ACCESSOR) ?? 0;
  const toolbarConfig = {
    ...toolbar,
    hidden: toolbar?.hidden || showLoading
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      "data-testid": "gauge-layout-container",
      ref: forwardedRef,
      className: import_GaugeChartLayout_css.layoutContainer,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_GaugeLayoutOverlay.GaugeChartOverlay,
        {
          value: displayValue,
          color: valueArcColor,
          intents,
          showLoading,
          toolbar: toolbarConfig,
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_GaugeRootSVG.GaugeRootSVG, { children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_GaugeBaseArc.GaugeBaseArc, {}),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_GaugeValueArc.GaugeValueArc,
              {
                value: displayValue,
                color: valueArcColor,
                min,
                max
              }
            ),
            !(0, import_lodash_es.isEmpty)(thresholds) && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_GaugeThresholdIndicators.GaugeThresholdIndicators, {}),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_GaugeLabels.GaugeLabels, { value: displayValue, prefixIcon })
          ] })
        }
      )
    }
  );
});
GaugeChartLayout.displayName = "GaugeChartLayout";
