// @dynatrace/strato-components/charts/gauge/components/GaugeChartLayout/GaugeChartToolbar.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var GaugeChartToolbar_exports = {};
__export(GaugeChartToolbar_exports, {
  GaugeChartToolbar: () => GaugeChartToolbar
});
module.exports = __toCommonJS(GaugeChartToolbar_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_useTooltipReducer = require("../../../core/components/overlay-tooltip/useTooltipReducer.js");
var import_ChartToolbarRenderer = require("../../../core/components/toolbar/ChartToolbarRenderer.js");
var import_zoom = require("../../../core/utils/zoom.js");
var import_constants = require("../../constants.js");
const GaugeChartToolbar = (props) => {
  const dispatch = (0, import_useTooltipReducer.useTooltipReducer)();
  const { intents, toolbar } = props;
  if (toolbar.hidden) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ChartToolbarRenderer.ChartToolbarRenderer,
    {
      visible: false,
      collapsed: false,
      draggable: true,
      placement: toolbar.placement,
      initialA11yState: (0, import_zoom.getA11yToolbarConfig)({
        intent: !!intents
      }),
      zIndex: import_constants.zIndexGaugeAreas.toolbar,
      intents,
      menuDisplayMode: "never",
      toolbarDisplayMode: "overlay",
      removeDragHandle: true,
      visibleWhenMouseInBounds: true,
      onToggleMenu: (isOpen) => {
        if (isOpen) {
          dispatch({ type: "RESET_TOOLTIP" })();
        }
      }
    }
  );
};
