// @dynatrace/strato-components/charts/meter-bar/single-meter-bar/MeterBarChart.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MeterBarChart_exports = {};
__export(MeterBarChart_exports, {
  MeterBarChart: () => MeterBarChart
});
module.exports = __toCommonJS(MeterBarChart_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_MeterBarChartPlotArea = require("./components/MeterBarChartPlotArea.js");
var import_MeterBarChartThresholdLegend = require("./components/MeterBarChartThresholdLegend/MeterBarChartThresholdLegend.js");
var import_constants = require("./constants.js");
var import_useMeterBarChartConfig = require("./hooks/useMeterBarChartConfig.js");
var import_MeterBarChartInternalDataProviders = require("./providers/MeterBarChartInternalDataProviders.js");
var import_MeterBarChartProviders = require("./providers/MeterBarChartProviders.js");
var import_Label = require("./slots/Label.js");
var import_Legend = require("./slots/Legend.js");
var import_Max = require("./slots/Max.js");
var import_Min = require("./slots/Min.js");
var import_ThresholdIndicator = require("./slots/ThresholdIndicator.js");
var import_ThresholdLegend = require("./slots/ThresholdLegend.js");
var import_Tooltip = require("./slots/Tooltip.js");
var import_Value = require("./slots/Value.js");
var import_useBehavioralTrackingProps = require("../../../core/hooks/useBehavioralTrackingProps.js");
var import_Intent = require("../../../core/slots/Intent/Intent.js");
var import_defaultChartValues_css = require("../../../core/styles/defaultChartValues.sty.js");
var import_constants2 = require("../../core/constants.js");
var import_ColorRule = require("../../core/slots/ColorRule.js");
var import_chart_types = require("../../core/types/chart-types.js");
var import_MeterBarChartFormattedValue = require("../shared/components/formatted-value/MeterBarChartFormattedValue.js");
var import_MeterBarLayout = require("../shared/components/meter-bar-layout/MeterBarLayout.js");
var import_Toolbar = require("../shared/slots/Toolbar.js");
var import_get_numeric_value = require("../shared/utils/get-numeric-value.js");
const COMPONENT_NAME = "SingleMeterBarChart";
const MeterBarChartRoot = (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    width,
    height,
    value,
    name,
    color,
    className,
    style,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "data-testid": dataTestId = "single-meter-bar-chart-root",
    prefixIcon,
    valueAccessor
  } = props;
  const [behavioralTrackingProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(props);
  const config = (0, import_useMeterBarChartConfig.useMeterBarChartConfig)(props);
  const {
    size,
    thresholds,
    thresholdsIndicators,
    loading,
    min,
    max,
    minSlot,
    maxSlot,
    valueSlot,
    labelSlot,
    tooltip,
    formatter,
    unit
  } = config;
  const meterBarChartRef = (0, import_react.useRef)(null);
  const hasLegend = thresholds.legend ?? thresholdsIndicators.legend;
  const consumerConfig = {
    type: import_chart_types.ChartType.SingleMeterBar,
    thresholdsIndicators,
    size,
    min,
    max,
    formatter,
    unit,
    tooltip: !tooltip.hidden,
    loading
  };
  const displayValue = (0, import_get_numeric_value.getNumericValue)(
    value,
    valueAccessor ?? import_constants2.DEFAULT_VALUE_ACCESSOR
  );
  const internalDataProvidersProps = {
    value,
    displayValue,
    valueAccessor,
    name,
    color
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_MeterBarChartProviders.MeterBarChartProviders, { config, seriesActions: props.seriesActions, children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_MeterBarLayout.MeterBarLayout,
    {
      ...behavioralTrackingProps,
      ref: meterBarChartRef,
      size,
      width,
      height,
      className: (0, import_clsx.default)(import_defaultChartValues_css.defaultChartSize, className),
      style,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "data-testid": dataTestId,
      "data-dt-component": COMPONENT_NAME,
      "aria-label": import_constants.METER_BAR_CHART_ARIA_LABEL,
      consumerConfig,
      userRef: forwardedRef,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_MeterBarLayout.MeterBarLayout.Providers,
          {
            providerName: import_MeterBarChartInternalDataProviders.MeterBarChartInternalDataProviders,
            providerProps: internalDataProvidersProps
          }
        ),
        prefixIcon && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_MeterBarLayout.MeterBarLayout.Icon, { children: prefixIcon }),
        labelSlot && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_MeterBarLayout.MeterBarLayout.Label, { children: labelSlot }),
        valueSlot && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_MeterBarLayout.MeterBarLayout.Value, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_MeterBarChartFormattedValue.MeterBarChartFormattedValue,
          {
            value: displayValue,
            total: max,
            auto: valueSlot?.auto,
            children: valueSlot?.value
          }
        ) }),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_MeterBarLayout.MeterBarLayout.Chart, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_MeterBarChartPlotArea.MeterBarChartPlotArea, {}) }),
        minSlot && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_MeterBarLayout.MeterBarLayout.Min, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_MeterBarChartFormattedValue.MeterBarChartFormattedValue, { value: min, auto: minSlot?.auto, children: minSlot?.value }) }),
        maxSlot && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_MeterBarLayout.MeterBarLayout.Max, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_MeterBarChartFormattedValue.MeterBarChartFormattedValue, { value: max, auto: maxSlot?.auto, children: maxSlot?.value }) }),
        hasLegend && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_MeterBarLayout.MeterBarLayout.Legend, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_MeterBarChartThresholdLegend.MeterBarChartThresholdLegend, {}) })
      ]
    }
  ) });
});
MeterBarChartRoot.displayName = "MeterBarChart";
const MeterBarChart = Object.assign(MeterBarChartRoot, {
  Value: import_Value.Value,
  Label: import_Label.Label,
  Min: import_Min.Min,
  Max: import_Max.Max,
  ThresholdIndicator: import_ThresholdIndicator.ThresholdIndicator,
  // eslint-disable-next-line @typescript-eslint/no-deprecated
  ThresholdLegend: import_ThresholdLegend.ThresholdLegend,
  Legend: import_Legend.Legend,
  Tooltip: import_Tooltip.Tooltip,
  ColorRule: import_ColorRule.ColorRule,
  Intent: import_Intent.Intent,
  Toolbar: import_Toolbar.Toolbar
});
