// @dynatrace/strato-components/esm/core/components/app-root/AppRoot.js
import { jsx } from "react/jsx-runtime";
import { isEmpty } from "lodash-es";
import {
  forwardRef,
  useEffect,
  useLayoutEffect,
  useState
} from "react";
import { IntlProvider } from "react-intl";
import "wicg-inert";
import { getLanguage, getTimezone } from "@dynatrace-sdk/user-preferences";
import * as styles from "./AppRoot.sty.js";
import { useCurrentTheme } from "../../hooks/useCurrentTheme.js";
import { FocusProvider } from "../../providers/FocusProvider.js";
import { isVisualRefreshEnabled } from "../../utils/is-visual-refresh-enabled.js";
import { initModalityDetection } from "../../utils/modality.js";
initModalityDetection();
const COMPONENT_NAME = "AppRoot";
function getBaseHref() {
  const baseElement = document.querySelector("base");
  const href = baseElement?.href ?? "/";
  return href.endsWith("/") ? href : `${href}/`;
}
const defaultLanguage = "en";
const AppRoot = /* @__PURE__ */ forwardRef(
  (props, forwardedRef) => {
    const {
      style: consumerStyle,
      className: consumerClassName,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      translationRootPath = `${getBaseHref()}lang/`
    } = props;
    const [messages, setMessages] = useState({});
    const { children } = props;
    const theme = useCurrentTheme();
    const language = getLanguage();
    let timezone = getTimezone();
    try {
      new Intl.DateTimeFormat(language, { timeZone: timezone });
    } catch {
      timezone = Intl.DateTimeFormat().resolvedOptions().timeZone;
    }
    useLayoutEffect(() => {
      if (document !== void 0) {
        document.documentElement.setAttribute("data-theme", theme);
      }
    }, [theme]);
    useLayoutEffect(() => {
      if (document !== void 0 && isVisualRefreshEnabled()) {
        document.documentElement.classList.add(styles.dtVisualRefreshEnabled);
      }
    }, []);
    useEffect(() => {
      if (language === "en") {
        return;
      }
      fetch(`${translationRootPath}${language}.json`).then((res) => {
        return res.json();
      }).then((receivedMessages) => {
        setMessages(receivedMessages);
      }).catch(() => {
        setMessages((prev) => isEmpty(prev) ? prev : {});
      });
    }, [language]);
    return /* @__PURE__ */ jsx(
      "div",
      {
        ref: forwardedRef,
        "data-testid": dataTestId,
        className: consumerClassName,
        style: consumerStyle,
        "data-theme": theme,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        "data-dt-component": COMPONENT_NAME,
        children: /* @__PURE__ */ jsx(
          IntlProvider,
          {
            locale: language,
            timeZone: timezone,
            messages,
            defaultLocale: defaultLanguage,
            children: /* @__PURE__ */ jsx(FocusProvider, { children })
          }
        )
      }
    );
  }
);
AppRoot.displayName = COMPONENT_NAME;
export {
  AppRoot
};
