// @dynatrace/strato-components/esm/charts/xy-chart/components/overlay/utils/chartCrosshairDataTransformer.js
import { units } from "@dynatrace-sdk/units";
import { getAxisFormatterWithTimeFallback } from "./get-axis-formatter-with-time-fallback.js";
import {
  getMagnifierLabel
} from "../../../../core/components/crosshair/utils/get-magnifier-label.js";
import { defaultUnitFormatter } from "../../../../core/utils/defaultUnitFormatter.js";
const coordinates = {
  top: "x",
  bottom: "x",
  left: "y",
  right: "y"
};
const defaultLabels = () => ({
  left: void 0,
  right: void 0,
  bottom: void 0,
  top: void 0
});
const accumulateFormatters = (getFormatter) => (acc, axis) => {
  const formatter = getFormatter(axis);
  if (formatter) {
    acc.push([axis, formatter]);
  }
  return acc;
};
const accumulateLabels = (getLabel) => (acc, axisWithFormatter) => {
  const [axis, formatter] = axisWithFormatter;
  const label = getLabel(axis, formatter);
  if (label) {
    acc[axis] = label;
  }
  return acc;
};
const chartCrosshairDataTransformer = ({
  axesScale,
  axesConfig,
  axesIdToFormatter,
  width,
  axes = ["left", "right", "bottom", "top"],
  formatCategory
}) => {
  const getFormatter = (axis) => {
    const axisConfig = axesConfig[axis];
    const scale = axesScale[axis];
    if (axisConfig) {
      if (axisConfig.isRelative) {
        return defaultUnitFormatter(units.percentage.percent);
      }
      return getAxisFormatterWithTimeFallback({
        axisConfig,
        axisId: axisConfig.id,
        scale,
        axesIdToFormatter,
        width
      });
    }
  };
  const formatters = axes.reduce(
    accumulateFormatters(getFormatter),
    []
  );
  return ({ position, pinned }) => {
    if (pinned) {
      return defaultLabels();
    }
    const getLabel = (axis, formatter) => {
      const raw = getMagnifierLabel(
        axesConfig[axis]?.unit,
        coordinates[axis],
        formatter,
        position,
        axesScale[axis]
      );
      if (raw === null || !formatCategory) {
        return raw;
      }
      return formatCategory(axis, raw) ?? raw;
    };
    return formatters.reduce(accumulateLabels(getLabel), defaultLabels());
  };
};
export {
  chartCrosshairDataTransformer
};
