// @dynatrace/strato-components/esm/charts/xy-chart/components/overlay/utils/get-selected-datapoints.js
import { isUndefined } from "lodash-es";
import { isRectangleSelected } from "./is-rectangle-selected.js";
import { serializeLinearSeriesValues } from "./serialize-linear-series-values.js";
import { serializeRectangleSeriesValues } from "./serialize-rectangle-series-values.js";
import { POINT_OVERLAY_PADDING } from "../../../../core/components/data-point/constants.js";
import { getSelectionPixels } from "../../../../core/utils/get-selection-pixels.js";
import { safeGet } from "../../../../core/utils/safe-get.js";
import { scaleBandInvert } from "../../../../core/utils/scale.js";
import {
  isAnyNumericalScale,
  isCategoricalScale,
  isDotContentPoint,
  isLineContentPoint,
  isAreaContentPoint
} from "../../../types/guards.js";
import {
  BAR_ID,
  BAR_VARIANT,
  DOT_ID,
  DOT_VARIANT,
  LINE_ID,
  LINE_VARIANT,
  RECT_ID,
  RECT_VARIANT,
  AREA_ID,
  AREA_VARIANT
} from "../../../types/xy-chart-internals.js";
import { getLinearAccessors } from "../../../utils/get-linear-accessors.js";
import {
  getCurrentScale,
  getDimensionScale
} from "../../../utils/get-scale.js";
import {
  isAreaSeries,
  isBarSeries,
  isDotSeries,
  isLineSeries,
  isRectSeries
} from "../../../utils/type-guards.js";
import { isGapDatapointForDisplay } from "../../../utils/validation-datapoint.js";
import {
  relAccessor
} from "../../../utils/value-representation.js";
import {
  x1Accessor,
  xAccessor,
  y1Accessor,
  yAccessor
} from "../../displays/area-accessors.js";
import {
  bandX0Accessor,
  bandX1Accessor,
  bandY0Accessor,
  bandY1Accessor
} from "../../displays/band-accessors.js";
import {
  heightAccessor as barHeightAccessor,
  widthAccessor as barWidthAccessor,
  xAccessor as xBarAccessor,
  yAccessor as yBarAccessor
} from "../../displays/bar-accessors.js";
import {
  isXBaselineAccessor,
  isYBaselineAccessor
} from "../../displays/BarDisplay/BarDisplay.js";
import {
  xAccessor as xDotAccessor,
  yAccessor as yDotAccessor
} from "../../displays/dot-accessors.js";
import {
  xAccessor as xLineAccessor,
  yAccessor as yLineAccessor
} from "../../displays/line-accessors.js";
import {
  heightAccessor,
  widthAccessor,
  xAccessor as xRectAccessor,
  yAccessor as yRectAccessor
} from "../../displays/rect-accessors.js";
function getCategoriesIncluded(categories, domainStart, domainEnd) {
  const startIndex = categories.indexOf(domainStart);
  const endIndex = categories.indexOf(domainEnd);
  return categories.slice(startIndex, endIndex);
}
function getSelectionDomain(selection, axisConfig) {
  if (isAnyNumericalScale(axisConfig)) {
    return [
      axisConfig.axisScale.invert(selection.start),
      axisConfig.axisScale.invert(selection.end)
    ];
  } else if (isCategoricalScale(axisConfig)) {
    const domainStart = scaleBandInvert(axisConfig.axisScale)(selection.start);
    const domainEnd = scaleBandInvert(axisConfig.axisScale)(selection.end);
    const categories = axisConfig.axisScale.domain();
    return getCategoriesIncluded(categories, domainStart, domainEnd);
  }
  return [selection.start, selection.end];
}
const getSelectedDatapoints = (seriesWithDatapoints, axisIdToScales, axes, selection, dotSeriesToMetadata, lineSeriesToMetadata, rectSeriesToMetadata, datapointMetadataMap, axisIdToScaleDomain, barSeriesToMetadata, areaSeriesToMetadata) => {
  const rectSeriesIndex = { current: 0 };
  const isDefined = isGapDatapointForDisplay(datapointMetadataMap);
  return seriesWithDatapoints.map((series) => {
    const xScale = getCurrentScale(axisIdToScales, series.xAxisId);
    const yScale = getCurrentScale(axisIdToScales, series.yAxisId);
    const [yMin, yMax] = yScale.range().toSorted((a, b) => a - b);
    const axisConfig = {
      axisType: axes.find((axis) => axis.id === series.xAxisId)?.type,
      axisScale: xScale
    };
    if (isDotSeries(series)) {
      const { xAxis, yAxis } = safeGet(dotSeriesToMetadata, series);
      const xRelAccessor = relAccessor(xAxis, axisIdToScaleDomain);
      const yRelAccessor = relAccessor(yAxis, axisIdToScaleDomain);
      return getDotSeriesSelection(
        series,
        dotSeriesToMetadata,
        xScale,
        yScale,
        selection,
        [yMin, yMax],
        axisConfig,
        xRelAccessor,
        yRelAccessor
      );
    } else if (isLineSeries(series)) {
      const { xAxis, yAxis } = safeGet(lineSeriesToMetadata, series);
      const xRelAccessor = relAccessor(xAxis, axisIdToScaleDomain);
      const yRelAccessor = relAccessor(yAxis, axisIdToScaleDomain);
      return getLineSeriesSelection(
        series,
        lineSeriesToMetadata,
        xScale,
        yScale,
        selection,
        [yMin, yMax],
        axisConfig,
        isDefined,
        xRelAccessor,
        yRelAccessor
      );
    } else if (isRectSeries(series)) {
      const { xAxis, yAxis } = safeGet(rectSeriesToMetadata, series);
      const xRelAccessor = relAccessor(xAxis, axisIdToScaleDomain);
      const yRelAccessor = relAccessor(yAxis, axisIdToScaleDomain);
      return getRectSeriesSelection(series, {
        rectSeriesIndex,
        xScale,
        yScale,
        selection,
        yRange: [yMin, yMax],
        axisConfig,
        isDefined,
        xRelAccessor,
        yRelAccessor
      });
    } else if (isBarSeries(series)) {
      const { xAxis, yAxis } = safeGet(barSeriesToMetadata, series);
      const xRelAccessor = relAccessor(xAxis, axisIdToScaleDomain);
      const yRelAccessor = relAccessor(yAxis, axisIdToScaleDomain);
      return getBarSeriesSelection(
        barSeriesToMetadata,
        series,
        selection,
        [yMin, yMax],
        axisConfig,
        datapointMetadataMap,
        axisIdToScales,
        xRelAccessor,
        yRelAccessor
      );
    } else if (isAreaSeries(series)) {
      const { xAxis, yAxis } = safeGet(areaSeriesToMetadata, series);
      const xRelAccessor = relAccessor(xAxis, axisIdToScaleDomain);
      const yRelAccessor = relAccessor(yAxis, axisIdToScaleDomain);
      return getAreaSeriesSelection(series, {
        areaSeriesToMetadata,
        xScale,
        yScale,
        selection,
        yRange: [yMin, yMax],
        axisConfig,
        isDefined,
        datapointMetadataMap,
        xRelAccessor,
        yRelAccessor
      });
    } else {
      return {
        seriesName: "",
        datapoints: [],
        domain: getSelectionDomain(selection, axisConfig)
      };
    }
  });
};
function getDotSeriesSelection(series, dotSeriesToMetadata, xScale, yScale, selection, yRange, axisConfig, xRelAccessor, yRelAccessor) {
  const filteredData = filterPointSeries(
    series.data,
    xDotAccessor(xScale, xRelAccessor),
    yDotAccessor(yScale, yRelAccessor),
    selection,
    yRange
  );
  return {
    seriesName: safeGet(dotSeriesToMetadata, series).name,
    datapoints: filteredData,
    domain: getSelectionDomain(selection, axisConfig)
  };
}
function getLineSeriesSelection(series, lineSeriesToMetadata, xScale, yScale, selection, yRange, axisConfig, isDefined, xRelAccessor, yRelAccessor) {
  const filteredData = filterPointSeries(
    series.data,
    xLineAccessor(xScale, xRelAccessor),
    yLineAccessor(yScale, yRelAccessor),
    selection,
    yRange,
    isDefined
  );
  return {
    seriesName: safeGet(lineSeriesToMetadata, series).name,
    datapoints: filteredData,
    domain: getSelectionDomain(selection, axisConfig)
  };
}
function isWithinVerticalDomain(yStart, yEnd, yMin, yMax) {
  const top = Math.min(yStart, yEnd);
  const bottom = Math.max(yStart, yEnd);
  return top >= yMin && bottom <= yMax;
}
function getRectSeriesSelection(series, ctx) {
  const {
    rectSeriesIndex,
    xScale,
    yScale,
    selection,
    yRange,
    axisConfig,
    isDefined,
    xRelAccessor,
    yRelAccessor
  } = ctx;
  const xAccessor2 = xRectAccessor(xScale, xRelAccessor);
  const yAccessor2 = yRectAccessor(yScale, yRelAccessor);
  const horizontalBandAccessor = widthAccessor(xScale, xRelAccessor);
  const verticalBandAccessor = heightAccessor(yScale, yRelAccessor);
  const [yMin, yMax] = yRange;
  const filteredDatapoints = series.filteredData.filter((datapoint) => {
    if (!isDefined(datapoint)) {
      return false;
    }
    const rectXStart = xAccessor2(datapoint);
    const rectXEnd = rectXStart + horizontalBandAccessor(datapoint);
    const rectYStart = yAccessor2(datapoint);
    const rectYEnd = rectYStart + verticalBandAccessor(datapoint);
    if (!isWithinVerticalDomain(rectYStart, rectYEnd, yMin, yMax)) {
      return false;
    }
    return isRectangleSelected({
      x: {
        rectangleStart: rectXStart,
        rectangleEnd: rectXEnd,
        selectionStart: selection.start,
        selectionEnd: selection.end
      },
      y: {
        rectangleStart: rectYStart,
        rectangleEnd: rectYEnd,
        selectionStart: yMin,
        selectionEnd: yMax
      }
    });
  });
  return {
    seriesName: `rect-series-${++rectSeriesIndex.current}`,
    datapoints: filteredDatapoints,
    domain: getSelectionDomain(selection, axisConfig)
  };
}
function getBarSeriesSelection(barSeriesToMetadata, series, selection, yRange, axisConfig, datapointMetadataMap, axisIdToScales, xRelAccessor, yRelAccessor) {
  const xScale = getCurrentScale(
    axisIdToScales,
    series.xAxisId,
    isXBaselineAccessor(series.data[0]) ? series.type : void 0
  );
  const yScale = getCurrentScale(
    axisIdToScales,
    series.yAxisId,
    isYBaselineAccessor(series.data[0]) ? series.type : void 0
  );
  const [yMin, yMax] = yRange;
  const dimensionXScale = getDimensionScale(axisIdToScales, series.xAxisId);
  const dimensionYScale = getDimensionScale(axisIdToScales, series.yAxisId);
  const xAccessor2 = xBarAccessor(
    xScale,
    datapointMetadataMap,
    dimensionXScale,
    series.id,
    xRelAccessor
  );
  const yAccessor2 = yBarAccessor(
    yScale,
    datapointMetadataMap,
    dimensionYScale,
    series.id,
    yRelAccessor
  );
  const horizontalBandAccessor = barWidthAccessor(
    xScale,
    datapointMetadataMap,
    dimensionXScale,
    xRelAccessor
  );
  const verticalBandAccessor = barHeightAccessor(
    yScale,
    datapointMetadataMap,
    dimensionYScale,
    yRelAccessor
  );
  const filteredData = series.data.filter((datapoint) => {
    const barXStart = xAccessor2(datapoint);
    const barXEnd = barXStart + horizontalBandAccessor(datapoint);
    const barYStart = yAccessor2(datapoint);
    const barYEnd = barYStart + verticalBandAccessor(datapoint);
    if (!isWithinVerticalDomain(barYStart, barYEnd, yMin, yMax)) {
      return false;
    }
    return isRectangleSelected({
      x: {
        rectangleStart: barXStart,
        rectangleEnd: barXEnd,
        selectionStart: selection.start,
        selectionEnd: selection.end
      },
      y: {
        rectangleStart: barYStart,
        rectangleEnd: barYEnd,
        selectionStart: yMin,
        selectionEnd: yMax
      }
    });
  });
  return {
    seriesName: safeGet(barSeriesToMetadata, series).name,
    datapoints: filteredData,
    domain: getSelectionDomain(selection, axisConfig)
  };
}
function getAreaSeriesSelection(series, ctx) {
  const {
    areaSeriesToMetadata,
    xScale,
    yScale,
    selection,
    yRange,
    axisConfig,
    isDefined,
    datapointMetadataMap,
    xRelAccessor,
    yRelAccessor
  } = ctx;
  const { name } = safeGet(areaSeriesToMetadata, series);
  const [yMin, yMax] = yRange;
  const { xAccessor: xAreaAccessor, yAccessor: yAreaAccessor } = getLinearAccessors(
    AREA_VARIANT,
    safeGet(areaSeriesToMetadata, series),
    datapointMetadataMap,
    xRelAccessor,
    yRelAccessor
  );
  const filteredData = series.data.filter((datapoint) => {
    if (!isDefined(datapoint)) {
      return false;
    }
    const xCoord = xAreaAccessor(xScale)(datapoint);
    const yCoord = yAreaAccessor(yScale)(datapoint);
    return yCoord >= yMin && yCoord <= yMax && xCoord >= selection.start && xCoord <= selection.end;
  });
  return {
    seriesName: name,
    datapoints: filteredData,
    domain: getSelectionDomain(selection, axisConfig)
  };
}
function filterPointSeries(data, xAccessor2, yAccessor2, selection, yRange, isDefined) {
  const [yMin, yMax] = yRange;
  return data.filter((datapoint) => {
    if (isDefined && !isDefined(datapoint)) {
      return false;
    }
    const pointX = xAccessor2(datapoint);
    const pointY = yAccessor2(datapoint);
    return pointX >= selection.start && pointX <= selection.end && pointY >= yMin && pointY <= yMax;
  });
}
function hasLinearContentPoint(contentState) {
  return !!contentState?.series && !!contentState?.datapoint && (contentState.type === DOT_VARIANT || contentState.type === LINE_VARIANT || contentState.type === AREA_VARIANT);
}
function collectLinearXAxisIds(dotSeries, lineSeries, areaSeries) {
  const ids = /* @__PURE__ */ new Set();
  for (const series of dotSeries) {
    if (series.xAxisId) {
      ids.add(series.xAxisId);
    }
  }
  for (const series of lineSeries) {
    if (series.xAxisId) {
      ids.add(series.xAxisId);
    }
  }
  for (const series of areaSeries) {
    if (series.xAxisId) {
      ids.add(series.xAxisId);
    }
  }
  return ids;
}
function getAxisIdToIsRelative(axes) {
  const axisIdToIsRelative = /* @__PURE__ */ new Map();
  axes.forEach((axis) => {
    axisIdToIsRelative.set(axis.id, axis.isRelative);
  });
  return axisIdToIsRelative;
}
function buildSelectionHighlightedPoints(params) {
  const {
    selectState,
    linearSelection,
    width,
    axisIdToScales,
    dotSeriesWithDatapoints,
    lineSeriesWithDatapoints,
    areaSeriesWithDatapoints,
    dotSeriesMeta,
    lineSeriesMeta,
    areaSeriesMeta,
    axisIdToScaleDomain,
    axes,
    datapointMetadataMap
  } = params;
  const anySelection = !selectState || linearSelection.selectionConfigByAxisId.size === 0;
  if (anySelection) {
    return [];
  }
  const axisIds = collectLinearXAxisIds(
    dotSeriesWithDatapoints,
    lineSeriesWithDatapoints,
    areaSeriesWithDatapoints
  );
  const axisIdToIsRelative = getAxisIdToIsRelative(axes);
  if (!axisIds.size) {
    return [];
  }
  const [selectionStart, selectionEnd] = getSelectionPixels(
    selectState.selectionPercentage,
    width,
    [selectState.start, selectState.end]
  );
  const { highlightedDatapoints } = linearSelection.searchAllAxisDatapoints(
    selectionStart.relative.x,
    selectionEnd.relative.x,
    serializeLinearSeriesValues(
      (series) => getLinearSeriesMetadata(
        series,
        dotSeriesMeta,
        lineSeriesMeta,
        areaSeriesMeta
      ),
      axisIdToScales,
      axisIdToScaleDomain,
      datapointMetadataMap
    ),
    Array.from(axisIds),
    axisIdToScales,
    axisIdToScaleDomain,
    axisIdToIsRelative
  );
  return highlightedDatapoints;
}
function buildDotHighlightedPoint(datapoint, meta, axisIdToScales, axisIdToScaleDomain) {
  const xScale = getCurrentScale(axisIdToScales, meta.xAxis.id);
  const yScale = getCurrentScale(axisIdToScales, meta.yAxis.id);
  const { xAxis, yAxis } = meta;
  const xRelAccessor = relAccessor(xAxis, axisIdToScaleDomain);
  const yRelAccessor = relAccessor(yAxis, axisIdToScaleDomain);
  return [
    {
      id: DOT_ID,
      x: xDotAccessor(xScale, xRelAccessor)(datapoint) + POINT_OVERLAY_PADDING,
      y: yDotAccessor(yScale, yRelAccessor)(datapoint) + POINT_OVERLAY_PADDING,
      color: meta.color,
      shape: meta.shape
    }
  ];
}
function buildLineHighlightedPoint(datapoint, meta, axisIdToScales, axisIdToScaleDomain) {
  const xScale = getCurrentScale(axisIdToScales, meta.xAxis.id);
  const yScale = getCurrentScale(axisIdToScales, meta.yAxis.id);
  const { xAxis, yAxis } = meta;
  const xRelAccessor = relAccessor(xAxis, axisIdToScaleDomain);
  const yRelAccessor = relAccessor(yAxis, axisIdToScaleDomain);
  return [
    {
      id: LINE_ID,
      x: xLineAccessor(xScale, xRelAccessor)(datapoint) + POINT_OVERLAY_PADDING,
      y: yLineAccessor(yScale, yRelAccessor)(datapoint) + POINT_OVERLAY_PADDING,
      color: meta.color,
      shape: void 0
    }
  ];
}
function buildHorizontalAreaPoints(datapoint, yScale, xScale, isBand, color, datapointMetadataMap, xRelAccessor, yRelAccessor, isXStacked) {
  const yCoord = isBand ? bandY0Accessor(yScale, yRelAccessor)(datapoint) + POINT_OVERLAY_PADDING : yAccessor(yScale, yRelAccessor)(datapoint) + POINT_OVERLAY_PADDING;
  const points = [];
  if (isBand) {
    points.push({
      id: AREA_ID,
      x: bandX0Accessor(xScale, xRelAccessor)(datapoint) + POINT_OVERLAY_PADDING,
      y: yCoord,
      color,
      shape: void 0
    });
  }
  points.push({
    id: AREA_ID,
    x: (isBand ? bandX1Accessor(xScale, xRelAccessor) : x1Accessor(xScale, datapointMetadataMap, xRelAccessor, isXStacked))(
      datapoint
    ) + POINT_OVERLAY_PADDING,
    y: yCoord,
    color,
    shape: void 0
  });
  return points;
}
function buildVerticalAreaPoints(datapoint, xScale, yScale, isBand, color, datapointMetadataMap, xRelAccessor, yRelAccessor, isYStacked) {
  const xCoord = isBand ? bandX0Accessor(xScale, xRelAccessor)(datapoint) + POINT_OVERLAY_PADDING : xAccessor(xScale, xRelAccessor)(datapoint) + POINT_OVERLAY_PADDING;
  const points = [];
  if (isBand) {
    points.push({
      id: AREA_ID,
      x: xCoord,
      y: bandY0Accessor(yScale, yRelAccessor)(datapoint) + POINT_OVERLAY_PADDING,
      color,
      shape: void 0
    });
  }
  points.push({
    id: AREA_ID,
    x: xCoord,
    y: (isBand ? bandY1Accessor(yScale, yRelAccessor) : y1Accessor(yScale, datapointMetadataMap, yRelAccessor, isYStacked))(
      datapoint
    ) + POINT_OVERLAY_PADDING,
    color,
    shape: void 0
  });
  return points;
}
function buildAreaHighlightedPoints(datapoint, meta, axisIdToScales, datapointMetadataMap, axisIdToScaleDomain) {
  const isHorizontal = meta.orientation === "horizontal";
  const isBand = meta.variant === "band";
  const xScale = getCurrentScale(axisIdToScales, meta.xAxis.id);
  const yScale = getCurrentScale(axisIdToScales, meta.yAxis.id);
  if (!xScale || !yScale) {
    return [];
  }
  const { xAxis, yAxis, isStacked } = meta;
  const xRelAccessor = relAccessor(xAxis, axisIdToScaleDomain);
  const yRelAccessor = relAccessor(yAxis, axisIdToScaleDomain);
  if (isHorizontal) {
    return buildHorizontalAreaPoints(
      datapoint,
      yScale,
      xScale,
      isBand,
      meta.color,
      datapointMetadataMap,
      xRelAccessor,
      yRelAccessor,
      isStacked
    );
  } else {
    return buildVerticalAreaPoints(
      datapoint,
      xScale,
      yScale,
      isBand,
      meta.color,
      datapointMetadataMap,
      xRelAccessor,
      yRelAccessor,
      isStacked
    );
  }
}
function buildContentHighlightedPoint(params) {
  const {
    contentState,
    axisIdToScales,
    dotSeriesMeta,
    lineSeriesMeta,
    areaSeriesMeta,
    axisIdToScaleDomain,
    datapointMetadataMap
  } = params;
  if (!hasLinearContentPoint(contentState)) {
    return [];
  }
  if (isDotContentPoint(contentState)) {
    const { series, datapoint } = contentState;
    const meta = safeGet(dotSeriesMeta, series);
    return buildDotHighlightedPoint(
      datapoint,
      meta,
      axisIdToScales,
      axisIdToScaleDomain
    );
  }
  if (isLineContentPoint(contentState)) {
    const { series, datapoint } = contentState;
    const meta = safeGet(lineSeriesMeta, series);
    return buildLineHighlightedPoint(
      datapoint,
      meta,
      axisIdToScales,
      axisIdToScaleDomain
    );
  }
  if (isAreaContentPoint(contentState)) {
    const { series, datapoint } = contentState;
    const meta = safeGet(areaSeriesMeta, series);
    return buildAreaHighlightedPoints(
      datapoint,
      meta,
      axisIdToScales,
      datapointMetadataMap,
      axisIdToScaleDomain
    );
  }
  return [];
}
function getLinearSeriesMetadata(series, dotSeriesMeta, lineSeriesMeta, areaSeriesMeta) {
  if (series.type === DOT_VARIANT) {
    return safeGet(dotSeriesMeta, series);
  }
  if (series.type === LINE_VARIANT) {
    return safeGet(lineSeriesMeta, series);
  }
  return safeGet(areaSeriesMeta, series);
}
function buildSelectionHighlightedRectangles(params) {
  const {
    selectState,
    rectSelection,
    width,
    rectSeriesMeta,
    rectSeriesWithDatapoints,
    barSeriesWithDatapoints,
    barSeriesMeta,
    axisIdToScales,
    axisIdToScaleDomain,
    axes,
    datapointMetadataMap
  } = params;
  if (!selectState || rectSelection.selectionConfigByAxisId.size === 0) {
    return [];
  }
  const definedAxes = Array.from(
    new Set(
      [...rectSeriesWithDatapoints, ...barSeriesWithDatapoints].flatMap(
        (series) => {
          const axes2 = [];
          if (!isUndefined(series.xAxisId)) {
            axes2.push(series.xAxisId);
          }
          if (!isUndefined(series.yAxisId)) {
            axes2.push(series.yAxisId);
          }
          return axes2;
        }
      )
    )
  );
  const axisIdToIsRelative = getAxisIdToIsRelative(axes);
  if (!definedAxes.length) {
    return [];
  }
  const [selectionStart, selectionEnd] = getSelectionPixels(
    selectState.selectionPercentage,
    width,
    [selectState.start, selectState.end]
  );
  const { highlightedDatapoints } = rectSelection.searchAllAxisDatapoints(
    selectionStart.relative.x,
    selectionEnd.relative.x,
    serializeRectangleSeriesValues(
      axisIdToScales,
      rectSeriesMeta,
      axisIdToScaleDomain,
      datapointMetadataMap,
      barSeriesMeta
    ),
    definedAxes,
    axisIdToScales,
    axisIdToScaleDomain,
    axisIdToIsRelative
  );
  return highlightedDatapoints;
}
function buildContentHighlightedRectangle(params) {
  const {
    contentState,
    axisIdToScales,
    barSeriesMeta,
    rectSeriesMeta,
    datapointMetadataMap,
    axisIdToScaleDomain
  } = params;
  if (!contentState?.series || !contentState?.datapoint || contentState.type !== BAR_VARIANT && contentState.type !== RECT_VARIANT) {
    return null;
  }
  if (contentState.type === BAR_VARIANT) {
    const { datapoint: datapoint2, series: series2 } = contentState;
    const { xAxis: xAxis2, yAxis: yAxis2 } = safeGet(barSeriesMeta, series2);
    const xScale2 = getCurrentScale(
      axisIdToScales,
      xAxis2.id,
      isXBaselineAccessor(series2.data[0]) ? series2.type : void 0
    );
    const dimensionXScale = getDimensionScale(axisIdToScales, series2.xAxisId);
    const yScale2 = getCurrentScale(
      axisIdToScales,
      yAxis2.id,
      isYBaselineAccessor(series2.data[0]) ? series2.type : void 0
    );
    const dimensionYScale = getDimensionScale(axisIdToScales, series2.yAxisId);
    const xRelAccessor2 = relAccessor(xAxis2, axisIdToScaleDomain);
    const yRelAccessor2 = relAccessor(yAxis2, axisIdToScaleDomain);
    return {
      id: BAR_ID,
      x: xBarAccessor(
        xScale2,
        datapointMetadataMap,
        dimensionXScale,
        series2.id,
        xRelAccessor2
      )(datapoint2),
      y: yBarAccessor(
        yScale2,
        datapointMetadataMap,
        dimensionYScale,
        series2.id,
        yRelAccessor2
      )(datapoint2),
      width: barWidthAccessor(
        xScale2,
        datapointMetadataMap,
        dimensionXScale,
        xRelAccessor2
      )(datapoint2),
      height: barHeightAccessor(
        yScale2,
        datapointMetadataMap,
        dimensionYScale,
        yRelAccessor2
      )(datapoint2)
    };
  }
  const { datapoint, series } = contentState;
  const { xAxis, yAxis } = safeGet(rectSeriesMeta, series);
  const xScale = getCurrentScale(axisIdToScales, xAxis.id);
  const yScale = getCurrentScale(axisIdToScales, yAxis.id);
  const xRelAccessor = relAccessor(xAxis, axisIdToScaleDomain);
  const yRelAccessor = relAccessor(yAxis, axisIdToScaleDomain);
  return {
    id: RECT_ID,
    x: xRectAccessor(xScale, xRelAccessor)(datapoint),
    y: yRectAccessor(yScale, yRelAccessor)(datapoint),
    width: widthAccessor(xScale, xRelAccessor)(datapoint),
    height: heightAccessor(yScale, yRelAccessor)(datapoint)
  };
}
function buildRectangleHighlightedPoints(params) {
  const selectionRects = buildSelectionHighlightedRectangles(params);
  const contentRect = buildContentHighlightedRectangle(params);
  return contentRect ? [...selectionRects, contentRect] : selectionRects;
}
export {
  buildContentHighlightedPoint,
  buildContentHighlightedRectangle,
  buildRectangleHighlightedPoints,
  buildSelectionHighlightedPoints,
  buildSelectionHighlightedRectangles,
  collectLinearXAxisIds,
  getSelectedDatapoints,
  hasLinearContentPoint
};
