// @dynatrace/strato-components/esm/charts/xy-chart/components/overlay/utils/serialize-linear-series-values.js
import { POINT_OVERLAY_PADDING } from "../../../../core/components/data-point/constants.js";
import { safeGet } from "../../../../core/utils/safe-get.js";
import {
  DOT_ID,
  DOT_VARIANT,
  LINE_ID,
  LINE_VARIANT,
  AREA_ID,
  AREA_VARIANT,
  BAND_VARIANT
} from "../../../types/xy-chart-internals.js";
import { getLinearAccessors } from "../../../utils/get-linear-accessors.js";
import { getCurrentScale } from "../../../utils/get-scale.js";
import { relAccessor } from "../../../utils/value-representation.js";
import {
  bandX0Accessor,
  bandX1Accessor,
  bandY0Accessor,
  bandY1Accessor
} from "../../displays/band-accessors.js";
function getYRangeMinMax(range) {
  let min = range[0];
  let max = range[0];
  for (let i = 1; i < range.length; i++) {
    const v = range[i];
    if (v < min) {
      min = v;
    } else if (v > max) {
      max = v;
    }
  }
  return min < max ? [min, max] : [max, min];
}
function resolveMetadata(getter, series) {
  if (typeof getter === "function") {
    return getter(series);
  }
  return safeGet(getter, series);
}
const idMap = {
  [DOT_VARIANT]: DOT_ID,
  [LINE_VARIANT]: LINE_ID,
  [AREA_VARIANT]: AREA_ID
};
function serializeBandArea(meta, point, xScale, yScale, xRelAcc, yRelAcc, yMin, yMax) {
  const coords = meta.orientation === "horizontal" ? [
    {
      x: bandX0Accessor(xScale, xRelAcc)(point.datapoint),
      y: bandY0Accessor(yScale, yRelAcc)(point.datapoint)
    },
    {
      x: bandX1Accessor(xScale, xRelAcc)(point.datapoint),
      y: bandY0Accessor(yScale, yRelAcc)(point.datapoint)
    }
  ] : [
    {
      x: bandX0Accessor(xScale, xRelAcc)(point.datapoint),
      y: bandY0Accessor(yScale, yRelAcc)(point.datapoint)
    },
    {
      x: bandX0Accessor(xScale, xRelAcc)(point.datapoint),
      y: bandY1Accessor(yScale, yRelAcc)(point.datapoint)
    }
  ];
  const visiblePoints = coords.filter(({ y }) => y >= yMin && y <= yMax).map(({ x, y }) => ({
    id: AREA_ID,
    x: x + POINT_OVERLAY_PADDING,
    y: y + POINT_OVERLAY_PADDING,
    color: meta.color,
    shape: void 0
  }));
  return visiblePoints.length > 0 ? visiblePoints : null;
}
const serializeLinearSeriesValues = (linearMetadata, axisIdToScales, axisIdToScaleDomain, datapointMetadataMap) => (point) => {
  const meta = resolveMetadata(linearMetadata, point.series);
  if (!meta) {
    return null;
  }
  const xScale = getCurrentScale(axisIdToScales, point.series.xAxisId);
  const yScale = getCurrentScale(axisIdToScales, point.series.yAxisId);
  if (!xScale || !yScale) {
    return null;
  }
  const { xAxis, yAxis } = meta;
  const xRelAccessor = relAccessor(xAxis, axisIdToScaleDomain);
  const yRelAccessor = relAccessor(yAxis, axisIdToScaleDomain);
  const { xAccessor, yAccessor } = getLinearAccessors(
    point.series.type,
    meta,
    datapointMetadataMap,
    xRelAccessor,
    yRelAccessor
  );
  const yBase = yAccessor(yScale)(point.datapoint);
  const [yMin, yMax] = getYRangeMinMax(yScale.range());
  if (yBase < yMin || yBase > yMax) {
    return null;
  }
  const areaMeta = point.series.type === AREA_VARIANT ? meta : void 0;
  if (areaMeta?.variant === BAND_VARIANT) {
    return serializeBandArea(
      areaMeta,
      point,
      xScale,
      yScale,
      xRelAccessor,
      yRelAccessor,
      yMin,
      yMax
    );
  }
  const xPos = xAccessor(xScale)(point.datapoint);
  return {
    id: idMap[point.series.type],
    x: xPos + POINT_OVERLAY_PADDING,
    y: yBase + POINT_OVERLAY_PADDING,
    color: meta.color,
    // shape is only defined for dot series
    shape: point.series.type === DOT_VARIANT ? meta.shape : void 0
  };
};
export {
  serializeLinearSeriesValues
};
