// @dynatrace/strato-components/esm/charts/xy-chart/components/overlay/utils/serialize-rectangle-series-values.js
import { safeGet } from "../../../../core/utils/safe-get.js";
import {
  BAR_ID,
  RECT_ID,
  RECT_VARIANT
} from "../../../types/xy-chart-internals.js";
import {
  getCurrentScale,
  getDimensionScale
} from "../../../utils/get-scale.js";
import { isRectDatapoint } from "../../../utils/type-guards.js";
import { relAccessor } from "../../../utils/value-representation.js";
import {
  heightAccessor as barHeightAccessor,
  widthAccessor as barWidthAccessor,
  xAccessor as barXAccessor,
  yAccessor as barYAccessor
} from "../../displays/bar-accessors.js";
import {
  isXBaselineAccessor,
  isYBaselineAccessor
} from "../../displays/BarDisplay/BarDisplay.js";
import {
  heightAccessor as rectHeightAccessor,
  widthAccessor as rectWidthAccessor,
  xAccessor as rectXAccessor,
  yAccessor as rectYAccessor
} from "../../displays/rect-accessors.js";
const serializeRectangleSeriesValues = (axisIdToScales, rectMetadata, axisIdToScaleDomain, datapointMetadataMap, barMetadata) => {
  return (point) => {
    if (isRectDatapoint(point.datapoint) && point.series.type === RECT_VARIANT) {
      const currentXScale = getCurrentScale(
        axisIdToScales,
        point.series.xAxisId
      );
      const currentYScale = getCurrentScale(
        axisIdToScales,
        point.series.yAxisId
      );
      const { xAxis, yAxis } = safeGet(rectMetadata, point.series);
      const xRelAccessor2 = relAccessor(xAxis, axisIdToScaleDomain);
      const yRelAccessor2 = relAccessor(yAxis, axisIdToScaleDomain);
      return {
        ...point,
        id: RECT_ID,
        x: rectXAccessor(currentXScale, xRelAccessor2)(point.datapoint),
        y: rectYAccessor(currentYScale, yRelAccessor2)(point.datapoint),
        width: rectWidthAccessor(currentXScale, xRelAccessor2)(point.datapoint),
        height: rectHeightAccessor(
          currentYScale,
          yRelAccessor2
        )(point.datapoint)
      };
    }
    const xScale = getCurrentScale(
      axisIdToScales,
      point.series.xAxisId,
      isXBaselineAccessor(point.series.data[0]) ? point.series.type : void 0
    );
    const dimensionXScale = getDimensionScale(
      axisIdToScales,
      point.series.xAxisId
    );
    const yScale = getCurrentScale(
      axisIdToScales,
      point.series.yAxisId,
      isYBaselineAccessor(point.series.data[0]) ? point.series.type : void 0
    );
    const dimensionYScale = getDimensionScale(
      axisIdToScales,
      point.series.yAxisId
    );
    const barMeta = barMetadata ? safeGet(barMetadata, point.series) : void 0;
    const xRelAccessor = barMeta ? relAccessor(barMeta.xAxis, axisIdToScaleDomain) : void 0;
    const yRelAccessor = barMeta ? relAccessor(barMeta.yAxis, axisIdToScaleDomain) : void 0;
    const x = barXAccessor(
      xScale,
      datapointMetadataMap,
      dimensionXScale,
      point.series.id,
      xRelAccessor
    )(point.datapoint);
    const y = barYAccessor(
      yScale,
      datapointMetadataMap,
      dimensionYScale,
      point.series.id,
      yRelAccessor
    )(point.datapoint);
    const width = barWidthAccessor(
      xScale,
      datapointMetadataMap,
      dimensionXScale,
      xRelAccessor
    )(point.datapoint);
    const height = barHeightAccessor(
      yScale,
      datapointMetadataMap,
      dimensionYScale,
      yRelAccessor
    )(point.datapoint);
    return {
      ...point,
      id: BAR_ID,
      x,
      y,
      width,
      height
    };
  };
};
export {
  serializeRectangleSeriesValues
};
