// @dynatrace/strato-components/charts/meter-bar/single-meter-bar/providers/MeterBarChartProviders.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MeterBarChartProviders_exports = {};
__export(MeterBarChartProviders_exports, {
  MeterBarChartProviders: () => MeterBarChartProviders
});
module.exports = __toCommonJS(MeterBarChartProviders_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_OverlayContext = require("../../../core/components/overlay-tooltip/contexts/OverlayContext.js");
var import_TooltipContext = require("../../../core/components/overlay-tooltip/contexts/TooltipContext.js");
var import_ChartSeriesActionContext = require("../../../core/components/series-action/ChartSeriesActionContext.js");
var import_ColorRuleContext = require("../../../core/context/ColorRuleContext.js");
var import_FormatterContext = require("../../shared/context/FormatterContext.js");
var import_MeterBarChartLoadingContext = require("../../shared/context/MeterBarChartLoadingContext.js");
var import_MeterBarToolbarConfigContext = require("../../shared/context/MeterBarToolbarConfigContext.js");
var import_MeterBarChartInternalConfigContext = require("../context/MeterBarChartInternalConfigContext.js");
const MeterBarChartProviders = ({
  config,
  seriesActions,
  children
}) => {
  const contextSeriesActionsValue = (0, import_react.useMemo)(
    () => ({
      chartSeriesActions: seriesActions
    }),
    [seriesActions]
  );
  const internalConfigValue = (0, import_react.useMemo)(
    () => ({
      min: config.min,
      max: config.max,
      tooltip: config.tooltip,
      label: config.labelSlot,
      thresholds: config.thresholds,
      thresholdsIndicators: config.thresholdsIndicators
    }),
    [config]
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ColorRuleContext.ColorRuleContext.Provider, { value: config.colorRules, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_FormatterContext.FormatterContext.Provider,
    {
      value: { formatter: config.formatter, unit: config.unit },
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_MeterBarChartLoadingContext.MeterBarChartLoadingContext.Provider, { value: config.loading, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ChartSeriesActionContext.ChartSeriesActionContext.Provider, { value: contextSeriesActionsValue, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_MeterBarChartInternalConfigContext.MeterBarChartInternalConfigContext.Provider,
        {
          value: internalConfigValue,
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_MeterBarToolbarConfigContext.MeterBarToolbarConfigContext.Provider, { value: config.toolbar, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayContext.OverlayProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TooltipContext.TooltipProvider, { children }) }) })
        }
      ) }) })
    }
  ) });
};
