// @dynatrace/strato-geo/esm/map/store/store.js
import { useContext } from "react";
import { MapStore } from "./map-store.context.js";
const useSetState = () => {
  const { setState } = useContext(MapStore);
  return setState;
};
export * from "./selectors.js";
const useSetStateSelection = () => {
  const setState = useSetState();
  return (newSelectionState) => {
    setState((prev) => ({
      ...prev,
      selection: {
        ...prev.selection,
        ...newSelectionState
      }
    }));
  };
};
const useSetStateOverlay = () => {
  const setState = useSetState();
  return (newOverlayState) => {
    setState((prev) => ({
      ...prev,
      overlay: {
        ...prev.overlay,
        ...newOverlayState
      }
    }));
  };
};
const useSetStateZoom = () => {
  const setState = useSetState();
  return (newZoomState) => {
    setState((prev) => ({
      ...prev,
      zoom: {
        ...prev.zoom,
        ...newZoomState
      }
    }));
  };
};
export {
  useSetState,
  useSetStateOverlay,
  useSetStateSelection,
  useSetStateZoom
};
