// @dynatrace/strato-components/filters/filter-field/decorations/decoration-marks.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var decoration_marks_exports = {};
__export(decoration_marks_exports, {
  errorMark: () => errorMark,
  escapedCharacterMark: () => escapedCharacterMark,
  existsMark: () => existsMark,
  incompleteStatementMark: () => incompleteStatementMark,
  intermediateStatementMark: () => intermediateStatementMark,
  keyMark: () => keyMark,
  pastingMark: () => pastingMark,
  postSpotErrorMark: () => postSpotErrorMark,
  preSpotErrorMark: () => preSpotErrorMark,
  searchMark: () => searchMark,
  statementMark: () => statementMark,
  valueMark: () => valueMark,
  variableMark: () => variableMark
});
module.exports = __toCommonJS(decoration_marks_exports);
var import_view = require("@codemirror/view");
var import_clsx = __toESM(require("clsx"));
var styles = __toESM(require("../FilterField.sty.js"));
const statementMark = import_view.Decoration.mark({
  inclusive: true,
  class: styles.filterStatement,
  tagName: "span"
});
const incompleteStatementMark = import_view.Decoration.mark({
  inclusive: true,
  class: styles.incompleteFilterStatement,
  tagName: "span"
});
const intermediateStatementMark = import_view.Decoration.mark({
  inclusive: true,
  class: styles.intermediateFilterStatement,
  tagName: "span"
});
const pastingMark = import_view.Decoration.mark({
  inclusive: true,
  class: styles.pastedContent,
  tagName: "span"
});
const existsMark = import_view.Decoration.mark({
  inclusive: true,
  class: styles.existsOperator,
  tagName: "span"
});
const searchMark = import_view.Decoration.mark({
  inclusive: true,
  class: styles.operator,
  tagName: "span"
});
const keyMark = import_view.Decoration.mark({
  inclusive: true,
  class: styles.filterKey,
  tagName: "span"
});
const valueMark = import_view.Decoration.mark({
  inclusive: true,
  class: styles.filterValue,
  tagName: "span"
});
const variableMark = import_view.Decoration.mark({
  inclusive: true,
  class: styles.variable,
  tagName: "span"
});
const escapedCharacterMark = import_view.Decoration.mark({
  inclusive: true,
  class: styles.escapedCharacter,
  tagName: "span"
});
const errorMark = import_view.Decoration.mark({
  inclusive: true,
  class: styles.error,
  tagName: "span"
});
const preSpotErrorMark = import_view.Decoration.mark({
  inclusive: true,
  class: (0, import_clsx.default)(styles.spotError, styles.preSpotError),
  tagName: "span"
});
const postSpotErrorMark = import_view.Decoration.mark({
  inclusive: true,
  class: (0, import_clsx.default)(styles.spotError, styles.postSpotError),
  tagName: "span"
});
