// @dynatrace/strato-components/filters/filter-field/decorations/filter-statements-delete-widget.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var filter_statements_delete_widget_exports = {};
__export(filter_statements_delete_widget_exports, {
  FilterStatementDeleteWidgetPlugin: () => FilterStatementDeleteWidgetPlugin,
  createFilterStatementDeleteWidgets: () => createFilterStatementDeleteWidgets
});
module.exports = __toCommonJS(filter_statements_delete_widget_exports);
var import_language = require("@codemirror/language");
var import_view = require("@codemirror/view");
var import_editor_config = require("../editor-state-configs/editor-config.js");
var import_FilterField_css = require("../FilterField.sty.js");
var import_useErrorLogsStore = require("../hooks/useErrorLogsStore.js");
var import_delete_widget_cleanup = require("../utils/delete-widget-cleanup.js");
var import_escape_suggestion = require("../utils/escape-suggestion.js");
var import_get_statement_status = require("../utils/get-statement-status.js");
var import_DeleteStatementWidget = require("../widgets/DeleteStatementWidget.js");
const needsEscapingRegex = new RegExp(import_escape_suggestion.NEEDS_ESCAPING_REGEX);
function createFilterStatementDeleteWidgets(view) {
  const widgets = [];
  const pastingStateField = view?.state.field(import_editor_config.pastingMode, false);
  const pastedContentHasSpecialChars = pastingStateField?.pastingInProgress && pastingStateField.range !== void 0 && needsEscapingRegex.test(
    view.state.doc.sliceString(
      pastingStateField.range.from,
      pastingStateField.range.to
    )
  );
  const pastedRange = pastedContentHasSpecialChars ? pastingStateField.range : void 0;
  for (const { from, to } of view.visibleRanges) {
    (0, import_language.syntaxTree)(view.state).iterate({
      from,
      to,
      enter: (node) => {
        if (node.name === "FilterStatement") {
          const { from: from2, to: to2 } = node;
          if (pastedRange && pastedRange.from <= to2 && pastedRange.to >= from2) {
            return false;
          }
          const statementType = (0, import_get_statement_status.getStatementStatus)(
            node.node.cursor(),
            view.state.doc.toString()
          );
          const deleteStatementWidget = import_view.Decoration.widget({
            widget: new import_DeleteStatementWidget.DeleteStatementWidget(
              from2,
              to2,
              statementType === "incomplete"
            )
          });
          widgets.push(deleteStatementWidget.range(to2));
          return false;
        }
      }
    });
  }
  return import_view.Decoration.set(widgets);
}
const FilterStatementDeleteWidgetPlugin = import_view.ViewPlugin.fromClass(
  class {
    decorations;
    constructor(view) {
      this.decorations = createFilterStatementDeleteWidgets(view);
    }
    update(update) {
      if (update.docChanged) {
        this.decorations = createFilterStatementDeleteWidgets(update.view);
      }
    }
  },
  {
    decorations: (view) => view.decorations,
    eventHandlers: {
      mousedown: (event, view) => {
        const target = event.target;
        if (target.nodeName === "BUTTON" && target.parentElement && target.parentElement.classList.contains(import_FilterField_css.filterStatementDeleteWidget)) {
          event.preventDefault();
          const setAction = import_useErrorLogsStore.useErrorLogsStore.getState().setAction;
          const { from: dataFrom, to: dataTo } = target.dataset;
          const parsedFrom = parseInt(dataFrom ?? "0");
          const parsedTo = parseInt(dataTo ?? "0");
          const currentValue = view.state.doc.toString();
          const adjustedRange = (0, import_delete_widget_cleanup.cleanUpInput)(view.state, {
            from: parsedFrom,
            to: parsedTo
          });
          const { from, to } = adjustedRange;
          setAction({
            type: "deleteStatement",
            data: {
              value: currentValue,
              range: { from, to },
              newCursorPos: from
            }
          });
          view.dispatch({
            changes: {
              from,
              to,
              insert: ""
            },
            annotations: import_editor_config.stringAnnotation.of(import_editor_config.ANNOTATION_DELETE_STATEMENT),
            selection: { anchor: from }
          });
        }
      }
    }
  }
);
