// @dynatrace/strato-components/filters/filter-field/decorations/filter-statements.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var filter_statements_exports = {};
__export(filter_statements_exports, {
  FilterStatementDecorationPlugin: () => FilterStatementDecorationPlugin,
  createFilterStatementDecorations: () => createFilterStatementDecorations
});
module.exports = __toCommonJS(filter_statements_exports);
var import_language = require("@codemirror/language");
var import_state = require("@codemirror/state");
var import_view = require("@codemirror/view");
var import_decoration_marks = require("./decoration-marks.js");
var import_editor_config = require("../editor-state-configs/editor-config.js");
var import_escape_suggestion = require("../utils/escape-suggestion.js");
var import_get_statement_status = require("../utils/get-statement-status.js");
const needsEscapingRegex = new RegExp(import_escape_suggestion.NEEDS_ESCAPING_REGEX);
function getStatementMark(statementType) {
  if (statementType === "complete") {
    return import_decoration_marks.statementMark;
  }
  if (statementType === "indeterminate") {
    return import_decoration_marks.intermediateStatementMark;
  }
  return import_decoration_marks.incompleteStatementMark;
}
function getMergedPasteRange(statementFrom, statementTo, pastedRange, currentMerged) {
  const rangeFrom = Math.min(statementFrom, pastedRange.from);
  const rangeTo = Math.max(statementTo, pastedRange.to);
  if (currentMerged === null) {
    return { from: rangeFrom, to: rangeTo };
  }
  return {
    from: Math.min(currentMerged.from, rangeFrom),
    to: Math.max(currentMerged.to, rangeTo)
  };
}
function isPasteAdjacentToStatement(pastedRange, statementFrom, statementTo, doc, statementType) {
  if (pastedRange.from <= statementTo && pastedRange.to >= statementFrom) {
    return true;
  }
  if (statementType === "complete" && pastedRange.from > statementTo) {
    return false;
  }
  const gap = pastedRange.from > statementTo ? doc.slice(statementTo, pastedRange.from) : doc.slice(pastedRange.to, statementFrom);
  return gap.trim() === "";
}
function collectRawDecorations(view, pastedRange) {
  const decorations = [];
  let mergedPasteRange = null;
  const doc = view.state.doc.toString();
  for (const { from, to } of view.visibleRanges) {
    (0, import_language.syntaxTree)(view.state).iterate({
      from,
      to,
      enter: (node) => {
        if (node.name === "FilterStatement") {
          const { from: statementFrom, to: statementTo } = node;
          const statementType = (0, import_get_statement_status.getStatementStatus)(node.node.cursor(), doc);
          const shouldMerge = pastedRange && (isPasteAdjacentToStatement(
            pastedRange,
            statementFrom,
            statementTo,
            doc,
            statementType
          ) || mergedPasteRange && isPasteAdjacentToStatement(
            mergedPasteRange,
            statementFrom,
            statementTo,
            doc,
            statementType
          ));
          if (shouldMerge) {
            mergedPasteRange = getMergedPasteRange(
              statementFrom,
              statementTo,
              pastedRange,
              mergedPasteRange
            );
          } else {
            decorations.push({
              from: statementFrom,
              to: statementTo,
              mark: getStatementMark(statementType),
              statementType
            });
          }
          return false;
        }
      }
    });
  }
  return { decorations, mergedPasteRange, doc };
}
function expandMergedRange(decorations, mergedPasteRange, doc) {
  const completeBoundaries = decorations.filter((dec) => dec.statementType === "complete").sort((a, b) => a.from - b.from);
  let leftBoundary = completeBoundaries.findLast((c) => c.to <= mergedPasteRange.from)?.to ?? 0;
  let rightBoundary = completeBoundaries.find((c) => c.from >= mergedPasteRange.to)?.from ?? doc.length;
  const leftGap = doc.slice(leftBoundary, mergedPasteRange.from);
  leftBoundary += leftGap.length - leftGap.trimStart().length;
  const rightGap = doc.slice(mergedPasteRange.to, rightBoundary);
  rightBoundary -= rightGap.length - rightGap.trimEnd().length;
  const expanded = {
    from: Math.min(mergedPasteRange.from, leftBoundary),
    to: Math.max(mergedPasteRange.to, rightBoundary)
  };
  const filtered = decorations.filter(
    (dec) => dec.statementType === "complete" || dec.to <= expanded.from || dec.from >= expanded.to
  );
  filtered.push({
    from: expanded.from,
    to: expanded.to,
    mark: import_decoration_marks.incompleteStatementMark,
    statementType: "incomplete"
  });
  return { decorations: filtered, mergedPasteRange: expanded };
}
function collectStatementDecorations(view, pastedRange) {
  const rawDecorations = collectRawDecorations(view, pastedRange);
  let { decorations, mergedPasteRange } = rawDecorations;
  const { doc } = rawDecorations;
  if (!mergedPasteRange && pastedRange) {
    mergedPasteRange = { from: pastedRange.from, to: pastedRange.to };
  }
  if (mergedPasteRange) {
    const expanded = expandMergedRange(decorations, mergedPasteRange, doc);
    decorations = expanded.decorations;
    mergedPasteRange = expanded.mergedPasteRange;
  }
  return { decorations, mergedPasteRange };
}
function createFilterStatementDecorations(view) {
  const builder = new import_state.RangeSetBuilder();
  const pastingStateField = view?.state.field(import_editor_config.pastingMode, false);
  const pastedContentHasSpecialChars = pastingStateField?.range !== void 0 && needsEscapingRegex.test(
    view.state.doc.sliceString(
      pastingStateField.range.from,
      pastingStateField.range.to
    )
  );
  const { decorations, mergedPasteRange } = collectStatementDecorations(
    view,
    pastedContentHasSpecialChars ? pastingStateField?.range : void 0
  );
  const allDecorations = decorations;
  if (pastedContentHasSpecialChars && pastingStateField?.pastingInProgress && pastingStateField.range) {
    const pastingMarkRange = mergedPasteRange ?? pastingStateField.range;
    for (const { from, to } of view.visibleRanges) {
      if (from <= pastingMarkRange.from && to >= pastingMarkRange.to) {
        allDecorations.push({
          from: pastingMarkRange.from,
          to: pastingMarkRange.to,
          mark: import_decoration_marks.pastingMark,
          statementType: "complete"
        });
      }
    }
  }
  allDecorations.sort((a, b) => a.from - b.from);
  for (const decoration of allDecorations) {
    builder.add(decoration.from, decoration.to, decoration.mark);
  }
  return builder.finish();
}
const FilterStatementDecorationPlugin = import_view.ViewPlugin.fromClass(
  class {
    decorations;
    constructor(view) {
      this.decorations = createFilterStatementDecorations(view);
    }
    update(update) {
      if (update.docChanged) {
        this.decorations = createFilterStatementDecorations(update.view);
      }
    }
  },
  {
    decorations: (view) => view.decorations
  }
);
