// @dynatrace/strato-components/filters/filter-field/decorations/highlight.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var highlight_exports = {};
__export(highlight_exports, {
  filterFieldHighlighter: () => filterFieldHighlighter,
  filterFieldStyleTags: () => filterFieldStyleTags
});
module.exports = __toCommonJS(highlight_exports);
var import_language = require("@codemirror/language");
var import_highlight = require("@lezer/highlight");
var import_FilterField_css = require("../FilterField.sty.js");
const filterFieldStyleTags = (0, import_highlight.styleTags)({
  ComparisonOperator: import_highlight.tags.compareOperator,
  ExistsOperator: import_highlight.tags.compareOperator,
  InclusionOperator: import_highlight.tags.compareOperator,
  LogicalOperator: import_highlight.tags.logicOperator
});
const filterFieldHighlighter = import_language.HighlightStyle.define([
  { tag: import_highlight.tags.compareOperator, class: import_FilterField_css.operator },
  { tag: import_highlight.tags.logicOperator, class: import_FilterField_css.operator }
]);
