// @dynatrace/strato-components/filters/timeframe-selector/constants/timeframeSelectorMessages.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var timeframeSelectorMessages_exports = {};
__export(timeframeSelectorMessages_exports, {
  TIMEFRAMESELECTOR_ERROR_MESSAGES: () => TIMEFRAMESELECTOR_ERROR_MESSAGES,
  TIMEFRAME_SELECTOR_MESSAGES: () => TIMEFRAME_SELECTOR_MESSAGES
});
module.exports = __toCommonJS(timeframeSelectorMessages_exports);
var import_react_intl = require("react-intl");
const TIMEFRAMESELECTOR_ERROR_MESSAGES = (0, import_react_intl.defineMessages)({
  minBoundary: {
    id: "KETnWaCwo9qLRQrI",
    defaultMessage: "Only dates after {min} are available.",
    description: "Given dates are not within the valid range."
  },
  maxBoundary: {
    id: "ONYnUtFidz4+6NmE",
    defaultMessage: "Only dates before {max} are available.",
    description: "Given dates are not within the valid range."
  },
  minMaxBoundary: {
    id: "Qm0/7EQKVXjS9tDM",
    defaultMessage: "Only dates between {min} \u2192 {max} are available.",
    description: "Given dates are not within the valid range."
  },
  invalidOrder: {
    defaultMessage: "'from' date should be before 'to' date",
    id: "DKFIWE64kHV1ohuk",
    description: "Message to display if the from and to dates are in the wrong order."
  },
  missingFrom: {
    defaultMessage: "Please enter a start time.",
    id: "GUeetpTpusqFkMi+",
    description: "From value is missing."
  },
  missingTo: {
    defaultMessage: "Please enter an end time.",
    id: "tfHTZkRc7sr/Ypcl",
    description: "End value is missing."
  },
  missingFromTo: {
    defaultMessage: "Please enter a start and end time.",
    id: "4yANH5gdmW1efk0I",
    description: "Start and end values are missing."
  },
  valueMissing: {
    defaultMessage: "Empty timeframe.",
    id: "QsKlSIxE+Yp58CUG",
    description: "Timeframe selector required timeframe error hint."
  },
  invalidFrom: {
    defaultMessage: "Invalid start time. Expected format: {expectedFormat}, -2h or -2d@d",
    id: "supuOfI2cSrNKOZ/",
    description: "Given from value is invalid"
  },
  invalidTo: {
    defaultMessage: "Invalid end time. Expected format: {expectedFormat}, -2h or -2d@d",
    id: "HL9dGWGnhYkm7OyC",
    description: "Given to value is invalid"
  },
  invalidFromTo: {
    defaultMessage: "Invalid start and end time. Expected format: {expectedFormat}, -2h or -2d@d",
    id: "89sBAyzLbeW6GyPJ",
    description: "Given from and to values are invalid"
  }
});
const TIMEFRAME_SELECTOR_MESSAGES = (0, import_react_intl.defineMessages)({
  startTime: {
    description: "Description for the start time of the preset item value, that is shown in the tooltip",
    defaultMessage: "Start",
    id: "QF4myF79Hrp3e/0Q"
  },
  endTime: {
    description: "Description for the end time of the preset item value, that is shown in the tooltip",
    defaultMessage: "End",
    id: "fHGjouAXIpSTl8T2"
  },
  duration: {
    description: "Description for the duration of the timeframe, that is shown in the tooltip",
    defaultMessage: "Duration",
    id: "ZvsC1dVcKeywqOWW"
  },
  prefix: {
    description: "Prefix for the moving a duration of the timeframe forward or backward, that is shown in the tooltip",
    defaultMessage: "Move ",
    id: "VoHb/9htaX3GgMyA"
  },
  suffix: {
    description: "Suffix for the moving a duration of the timeframe forward or backward, that is shown in the tooltip",
    defaultMessage: " {suffix}",
    id: "IU+dkogAuDAabi52"
  },
  "ms-duration": {
    description: "Timeframe duration formatter: timeframe-millisecond",
    defaultMessage: "{offset, plural, one {# millisecond} other {# milliseconds}} ",
    id: "L6ZeI3XBOuh0L/CT"
  },
  "s-duration": {
    description: "Timeframe duration formatter: timeframe-second",
    defaultMessage: "{offset, plural, one {# second} other {# seconds}} ",
    id: "mG+nMDkU8bB8zLlO"
  },
  "m-duration": {
    description: "Timeframe duration formatter: timeframe-minute",
    defaultMessage: "{offset, plural, one {# minute} other {# minutes}} ",
    id: "JlWcGe/hRo/kyTD+"
  },
  "h-duration": {
    description: "Timeframe duration formatter: timeframe-hour",
    defaultMessage: "{offset, plural, one {# hour} other {# hours}}",
    id: "ADLDZ8spJCrDx/6Q"
  },
  "d-duration": {
    description: "Timeframe duration formatter: timeframe-day",
    defaultMessage: "{offset, plural, one {# day} other {# days}}",
    id: "gIQd1giAq7WNDZp6"
  },
  "w-duration": {
    description: "Timeframe duration formatter: timeframe-week",
    defaultMessage: "{offset, plural, one {# week} other {# weeks}}",
    id: "TuHiu5OflYhkrvKy"
  },
  "M-duration": {
    description: "Timeframe duration formatter: timeframe-month",
    defaultMessage: "{offset, plural, one {# month} other {# months}} ",
    id: "AYfNQEjfwenkbxsg"
  },
  "q-duration": {
    description: "Timeframe duration formatter: timeframe-quarter",
    defaultMessage: "{offset, plural, one {# quarter} other {# quarters}}",
    id: "ItflQisz88wfEcF+"
  },
  "y-duration": {
    description: "Timeframe duration formatter: timeframe-year",
    defaultMessage: "{offset, plural, one {# year} other {# years}}",
    id: "gzgHXY7+xENWvOfC"
  },
  absoluteDateDuration: {
    description: "Timeframe duration formatter: move-absolute-date",
    defaultMessage: "{years, select, 0 {} other {{years, plural, one {# year } other {# years }}}}{months, select, 0 {} other {{months, plural, one {# month } other {# months }}}}{days, select, 0 {} other {{days, plural, one {# day } other {# days }}}}{hours, select, 0 {} other {{hours, plural, one {# hour } other {# hours }}}}{minutes, select, 0 {} other {{minutes, plural, one {# minute } other {# minutes }}}}{seconds, select, 0 {} other {{seconds, plural, one {# second } other {# seconds }}}}{milliseconds, select, 0 {} other {{milliseconds, plural, one {# millisecond } other {# milliseconds }}}}",
    id: "sITxaoPwIvMbTOY3"
  },
  docsLinkText: {
    defaultMessage: "Learn more",
    id: "q9YfvvFrRo1n8wvN",
    description: "Display text for the link to the timeframe selector documentation."
  },
  resetTimeframeButton: {
    defaultMessage: "Reset timeframe",
    id: "EG+qZn4zIajJ41CP",
    description: "Shown for the reset timeframe button of the timeframe selector."
  },
  timeframePresets: {
    defaultMessage: "List of timeframe presets",
    id: "1cycyeSJyv4MEQyz",
    description: "Aria label for the list of timeframe presets."
  },
  recentTimeframesHeading: {
    defaultMessage: "Recently used",
    id: "8kqNkvIZtJIon5xR",
    description: "Heading for the recently used timeframes."
  },
  recentTimeframesList: {
    defaultMessage: "List of recently used timeframes",
    id: "BDzwWccOiezgFJp3",
    description: "Aria label for the list of recently used timeframes."
  },
  disabledTooltip: {
    defaultMessage: "The timeframe exceeds the limits supported in this app.",
    id: "KzBG7IQ4sG79Dlpy",
    description: "Tooltip message for a disabled timeframe."
  },
  overlayContainer: {
    defaultMessage: "Timeframe selector overlay for providing a custom timeframe or selecting one of the presets",
    id: "hnsZ5MLEj9V3vVxP",
    description: "Aria label for the overlay of the timeframe selector."
  },
  applyButton: {
    defaultMessage: "Apply",
    id: "Np6+N27mrjvIJEls",
    description: "Shown for the apply button of the timeframe selector."
  },
  timeframeLabel: {
    defaultMessage: "Timeframe",
    id: "EKDlkAM4+D7rcALf",
    description: "Timeframe label shown in the TimeframeSelector overlay."
  }
});
