// @dynatrace/strato-components/esm/tables/DataTable/components/Toolbar/ColumnSettings/ColumnPinningMenu.js
import { jsx, jsxs } from "react/jsx-runtime";
import { FormattedMessage, useIntl } from "react-intl";
import { PinIcon } from "@dynatrace/strato-icons";
import { Button } from "../../../../../buttons/button/Button.js";
import { Menu } from "../../../../../navigation/menu/Menu.js";
import { Tooltip } from "../../../../../overlays/tooltip/Tooltip.js";
import { pinColumns } from "../../../features/ColumnPinning/column-pinning-utils.js";
import { dataTableMessages } from "../../../messages.js";
const ColumnPinningMenu = (props) => {
  const {
    column,
    orderedFlatColumns,
    localColumnPinning,
    updateLocalColumnPinning
  } = props;
  const intl = useIntl();
  const columnId = column.id;
  const pinnedPosition = column.pinned;
  const pinningLabel = intl.formatMessage(dataTableMessages.pinColumnLabel, {
    isUnpinned: pinnedPosition === false
  });
  function getColumnsInSamePinningRegion() {
    const position = pinnedPosition === false ? "center" : pinnedPosition;
    return orderedFlatColumns[position];
  }
  function getIdsToPin() {
    const hasNestedColumns = column.columns.length;
    if (!hasNestedColumns) {
      return [columnId];
    }
    const idsToPin = [];
    const columns = getColumnsInSamePinningRegion();
    const startIndex = columns.findIndex((col) => col.key === column.key);
    if (startIndex < 0) {
      return [];
    }
    let index = startIndex + 1;
    let current = columns.at(index);
    while (current?.parent?.id === columnId) {
      idsToPin.push(current.id);
      index++;
      current = columns.at(index);
    }
    return idsToPin;
  }
  const pin = (position) => {
    const idsToPin = getIdsToPin();
    const newState = pinColumns(idsToPin, localColumnPinning, position);
    updateLocalColumnPinning(newState);
  };
  return /* @__PURE__ */ jsx(Tooltip, { text: pinningLabel, placement: "right", children: /* @__PURE__ */ jsxs(Menu, { children: [
    /* @__PURE__ */ jsx(Menu.Trigger, { children: /* @__PURE__ */ jsx(Button, { size: "condensed", "aria-label": pinningLabel, children: /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(PinIcon, {}) }) }) }),
    /* @__PURE__ */ jsxs(Menu.Content, { children: [
      pinnedPosition !== "left" && /* @__PURE__ */ jsx(
        Menu.Item,
        {
          onSelect: () => pin("left"),
          "data-dt-name": "ColumnSettingsPinLeft",
          children: /* @__PURE__ */ jsx(FormattedMessage, { ...dataTableMessages.pinColumnLeftLabel })
        }
      ),
      pinnedPosition !== "right" && /* @__PURE__ */ jsx(
        Menu.Item,
        {
          onSelect: () => pin("right"),
          "data-dt-name": "ColumnSettingsPinRight",
          children: /* @__PURE__ */ jsx(FormattedMessage, { ...dataTableMessages.pinColumnRightLabel })
        }
      ),
      pinnedPosition !== false && /* @__PURE__ */ jsx(
        Menu.Item,
        {
          onSelect: () => pin(false),
          "data-dt-name": "ColumnSettingsUnpin",
          children: /* @__PURE__ */ jsx(FormattedMessage, { ...dataTableMessages.unpinColumnLabel })
        }
      )
    ] })
  ] }) });
};
export {
  ColumnPinningMenu
};
