// @dynatrace/strato-components/esm/tables/DataTable/components/Toolbar/ColumnSettings/DataTableColumnsListItem.js
import { jsx, jsxs } from "react/jsx-runtime";
import { useSortable } from "@dnd-kit/sortable";
import clsx from "clsx";
import { memo, useRef } from "react";
import { ColumnOrderDragHandle } from "./ColumnOrderSettings/ColumnOrderDragHandle.js";
import { ColumnPinningMenu } from "./ColumnPinningMenu.js";
import { ColumnVisibilityCheckbox } from "./ColumnVisibilityCheckbox.js";
import * as styles from "./DataTableColumnsListItem.sty.js";
import { useMergeRefs } from "../../../../../core/hooks/useMergeRefs.js";
import { Flex } from "../../../../../layouts/flex/Flex.js";
import { getColumnLabel } from "../../../../utils/get-column-label.js";
const InternalDataTableColumnsListItem = (props) => {
  const {
    column,
    columnIndex,
    orderedFlatColumns,
    showVisibilityCheckbox,
    enableVisibilityCheckbox,
    currentVisibilityState,
    onVisibilityChange,
    showColumnOrderDragHandle,
    showColumnPinningMenu,
    localColumnPinning,
    updateLocalColumnPinning,
    virtualIndex,
    measureElement
  } = props;
  const columnName = getColumnLabel(column);
  const localColumnRef = useRef(null);
  const { setNodeRef, isDragging } = useSortable({
    id: column.key,
    data: {
      column,
      columnId: column.id,
      columnIndex,
      localColumnRef
    }
  });
  const mergedRefs = useMergeRefs([localColumnRef, setNodeRef, measureElement]);
  return /* @__PURE__ */ jsxs(
    Flex,
    {
      "data-index": virtualIndex,
      "data-columnid": column.id,
      "data-parentcolumnid": column.parent?.id,
      "data-pinned-position": column.pinned === false ? void 0 : column.pinned,
      ref: mergedRefs,
      justifyContent: "space-between",
      alignItems: "center",
      className: clsx("strato-table-settings-item", styles.columnsListItem, {
        [styles.dragging]: isDragging
      }),
      style: {
        gridRowStart: virtualIndex + 1
      },
      children: [
        showVisibilityCheckbox ? /* @__PURE__ */ jsx(
          ColumnVisibilityCheckbox,
          {
            column,
            enableCheckbox: enableVisibilityCheckbox,
            currentState: currentVisibilityState,
            onVisibilityChange
          }
        ) : /* @__PURE__ */ jsx("div", { className: styles.columnName, children: columnName }),
        /* @__PURE__ */ jsxs(Flex, { gap: 4, children: [
          showColumnOrderDragHandle && /* @__PURE__ */ jsx(
            ColumnOrderDragHandle,
            {
              columnId: column.id,
              columnKey: column.key,
              isDraggingActive: isDragging
            }
          ),
          showColumnPinningMenu && /* @__PURE__ */ jsx(
            ColumnPinningMenu,
            {
              column,
              orderedFlatColumns,
              localColumnPinning,
              updateLocalColumnPinning
            }
          )
        ] })
      ]
    }
  );
};
const DataTableColumnsListItem = memo(
  InternalDataTableColumnsListItem,
  (prevProps, nextProps) => {
    return prevProps.column.key === nextProps.column.key && prevProps.columnIndex === nextProps.columnIndex && prevProps.showVisibilityCheckbox === nextProps.showVisibilityCheckbox && prevProps.enableVisibilityCheckbox === nextProps.enableVisibilityCheckbox && prevProps.currentVisibilityState === nextProps.currentVisibilityState && prevProps.showColumnOrderDragHandle === nextProps.showColumnOrderDragHandle && prevProps.showColumnPinningMenu === nextProps.showColumnPinningMenu && prevProps.localColumnPinning === nextProps.localColumnPinning && prevProps.virtualIndex === nextProps.virtualIndex;
  }
);
export {
  DataTableColumnsListItem
};
