// @dynatrace/strato-components/esm/tables/DataTable/components/Toolbar/ColumnSettings/DataTableVisibilityFilterControls.js
import { jsx, jsxs } from "react/jsx-runtime";
import { debounce } from "lodash-es";
import {
  useCallback,
  useEffect,
  useRef,
  useState
} from "react";
import { useIntl } from "react-intl";
import * as styles from "./DataTableVisibilityFilterControls.sty.js";
import { Checkbox } from "../../../../../forms/checkbox/Checkbox.js";
import { SearchInput } from "../../../../../forms/search-input/SearchInput.js";
import { Flex } from "../../../../../layouts/flex/Flex.js";
import { dataTableMessages } from "../../../messages.js";
const SEARCH_DEBOUNCE_TIME_MS = 300;
const DataTableVisibilityFilterControls = (props) => {
  const {
    setActiveFilterValue,
    showHideAllState,
    onShowHideAllChange,
    showVisibilityControl,
    showFilterControl
  } = props;
  const intl = useIntl();
  const [filterValue, setFilterValue] = useState("");
  const debouncedUpdateRef = useRef(
    debounce((value, setter) => {
      setter(value);
    }, SEARCH_DEBOUNCE_TIME_MS)
  );
  const onChange = useCallback(
    (value) => {
      setFilterValue(value);
      debouncedUpdateRef.current(value, setActiveFilterValue);
    },
    [setActiveFilterValue]
  );
  useEffect(() => {
    const debouncedUpdate = debouncedUpdateRef.current;
    return () => {
      debouncedUpdate.cancel();
    };
  }, []);
  if (!showFilterControl && !showVisibilityControl) {
    return null;
  }
  const dataDtProperties = `state:${showHideAllState === true ? "hide" : "show"}`;
  return /* @__PURE__ */ jsxs(
    Flex,
    {
      justifyContent: "space-between",
      alignItems: "center",
      className: styles.filterControls,
      children: [
        showVisibilityControl && /* @__PURE__ */ jsx(
          Checkbox,
          {
            value: showHideAllState,
            onChange: onShowHideAllChange,
            className: styles.showHideAll,
            "data-dt-name": "ColumnSettingsShowHideAll",
            "data-dt-properties": dataDtProperties,
            children: intl.formatMessage(dataTableMessages.showHideAll)
          }
        ),
        showFilterControl && /* @__PURE__ */ jsx(
          SearchInput,
          {
            placeholder: intl.formatMessage(
              dataTableMessages.searchBarPlaceholder
            ),
            className: styles.searchInput,
            onChange,
            value: filterValue,
            "data-dt-name": "ColumnSettingsFilter"
          }
        )
      ]
    }
  );
};
export {
  DataTableVisibilityFilterControls
};
