// @dynatrace/strato-components/esm/tables/DataTable/components/Toolbar/ColumnSettings/DataTableColumnsList.js
import { jsx } from "react/jsx-runtime";
import { useVirtualizer } from "@tanstack/react-virtual";
import clsx from "clsx";
import { useCallback, useMemo, useRef } from "react";
import { useIntl } from "react-intl";
import {
  generateColumnLabelsMap,
  getVisibilityCheckboxState,
  groupColumns,
  isParentColumnHideable,
  processFilterValue
} from "./column-settings-utils.js";
import { DATA_TABLE_COLUMNS_LIST_ITEM_MIN_HEIGHT } from "./ColumnOrderSettings/column-order-settings-dnd-utils.js";
import { ColumnOrderSettingsDragAndDropContext } from "./ColumnOrderSettings/ColumnOrderSettingsDragAndDrop.js";
import * as styles from "./DataTableColumnsList.sty.js";
import { columnName } from "./DataTableColumnsListItem.sty.js";
import { DataTableColumnsListItem } from "./DataTableColumnsListItem.js";
import { Grid } from "../../../../../layouts/grid/Grid.js";
import { Text } from "../../../../../typography/text/Text.js";
import { useDataTable } from "../../../contexts/DataTableContext.js";
import { getCanHide } from "../../../features/ColumnVisibility/ColumnVisibility.js";
import { dataTableMessages } from "../../../messages.js";
const DataTableColumnsList = (props) => {
  const {
    visibilitySettings: showVisibilityCheckbox,
    columnOrderSettings,
    columnPinningSettings,
    activeFilterValue,
    localColumnVisibility,
    updateLocalColumnVisibility,
    localColumnOrder,
    updateLocalColumnOrder,
    localColumnPinning,
    updateLocalColumnPinning,
    hasFilterBar
  } = props;
  const table = useDataTable();
  const intl = useIntl();
  const parentRef = useRef(null);
  const allLeafColumnsWithoutBuiltin = table.getAllLeafColumnsWithoutBuiltin();
  const columnsById = useMemo(
    () => new Map(allLeafColumnsWithoutBuiltin.map((col) => [col.id, col])),
    [allLeafColumnsWithoutBuiltin]
  );
  const columnLabelsById = useMemo(
    () => generateColumnLabelsMap(allLeafColumnsWithoutBuiltin),
    [allLeafColumnsWithoutBuiltin]
  );
  const activeFilterWords = useMemo(
    () => processFilterValue(activeFilterValue),
    [activeFilterValue]
  );
  const compareColumns = useCallback(
    (a, b) => {
      const idChildColumnA = a.columns.at(0)?.id;
      const idChildColumnB = b.columns.at(0)?.id;
      const isBChildOfA = !!a.columns.find((childCol) => childCol.id === b.id);
      const isAChildOfB = !!b.columns.find((childCol) => childCol.id === a.id);
      if (isBChildOfA) {
        return -1;
      } else if (isAChildOfB) {
        return 1;
      }
      return localColumnOrder.indexOf(idChildColumnA ?? a.id) - localColumnOrder.indexOf(idChildColumnB ?? b.id);
    },
    [localColumnOrder]
  );
  const filterColumns = useCallback(
    (column, _, columnSection) => {
      if (activeFilterWords.length === 0) {
        return true;
      }
      const matchesFilter = (columnId) => {
        const columnLabel = columnLabelsById.get(columnId) ?? "";
        return activeFilterWords.every((word) => columnLabel.includes(word));
      };
      const childColumns = column.columns;
      if (childColumns.length === 0) {
        return matchesFilter(column.id);
      }
      return childColumns.filter((child) => columnSection.find((col) => col.id === child.id)).some((childColumn) => matchesFilter(childColumn.id));
    },
    [activeFilterWords, columnLabelsById]
  );
  const mapToListItem = (col, idx, pinnedPosition) => ({
    ...col,
    key: `${col.id}_${idx}_${pinnedPosition}`,
    // unique React key; parent columns may appear multiple times in list
    pinned: pinnedPosition
  });
  const orderedLeftFlatColumns = useMemo(() => {
    const localColumnsPinned = localColumnPinning.left?.map((id) => columnsById.get(id)).filter((column) => !!column) ?? [];
    const groupedColumns = groupColumns(localColumnsPinned);
    return groupedColumns.map((col, idx) => mapToListItem(col, idx, "left"));
  }, [columnsById, localColumnPinning.left]);
  const orderedCenterFlatColumns = useMemo(() => {
    const localColumnsUnpinned = allLeafColumnsWithoutBuiltin.filter(
      (column) => !localColumnPinning.left?.includes(column.id) && !localColumnPinning.right?.includes(column.id)
    );
    const groupedColumns = groupColumns(localColumnsUnpinned);
    return groupedColumns.toSorted(compareColumns).map((col, idx) => mapToListItem(col, idx, false));
  }, [
    allLeafColumnsWithoutBuiltin,
    compareColumns,
    localColumnPinning.left,
    localColumnPinning.right
  ]);
  const orderedRightFlatColumns = useMemo(() => {
    const localColumnsPinned = localColumnPinning.right?.map((id) => columnsById.get(id)).filter((column) => !!column) ?? [];
    const groupedColumns = groupColumns(localColumnsPinned);
    return groupedColumns.map((col, idx) => mapToListItem(col, idx, "right"));
  }, [columnsById, localColumnPinning.right]);
  const orderedFlatColumns = {
    left: orderedLeftFlatColumns,
    center: orderedCenterFlatColumns,
    right: orderedRightFlatColumns
  };
  const filteredColumns = useMemo(
    () => [
      ...orderedLeftFlatColumns.filter(filterColumns),
      ...orderedCenterFlatColumns.filter(filterColumns),
      ...orderedRightFlatColumns.filter(filterColumns)
    ],
    [
      filterColumns,
      orderedCenterFlatColumns,
      orderedLeftFlatColumns,
      orderedRightFlatColumns
    ]
  );
  const getItemKey = useCallback(
    (index) => filteredColumns[index].key,
    [filteredColumns]
  );
  const columnVirtualizer = useVirtualizer({
    count: filteredColumns.length,
    overscan: 5,
    getScrollElement: () => parentRef.current,
    getItemKey,
    // Approximate height of a column list item
    estimateSize: () => DATA_TABLE_COLUMNS_LIST_ITEM_MIN_HEIGHT,
    measureElement: (element) => {
      if (!element || !(element instanceof HTMLElement)) {
        return DATA_TABLE_COLUMNS_LIST_ITEM_MIN_HEIGHT;
      }
      const labelElement = element.querySelector(`.${columnName}`);
      if (!labelElement) {
        return DATA_TABLE_COLUMNS_LIST_ITEM_MIN_HEIGHT;
      }
      const computedStyle = getComputedStyle(element);
      const borderTop = parseFloat(computedStyle.borderTopWidth);
      const borderBottom = parseFloat(computedStyle.borderBottomWidth);
      return labelElement.getBoundingClientRect().height + 2 * 4 + // padding top and bottom
      borderTop + borderBottom;
    }
  });
  const virtualItems = columnVirtualizer.getVirtualItems();
  if (filteredColumns.length === 0) {
    return /* @__PURE__ */ jsx(Text, { className: styles.noItemsFound, children: intl.formatMessage(dataTableMessages.noItemsFound) });
  }
  const showColumnPinningButton = !!table.options.enableColumnPinning && columnPinningSettings;
  const canOrder = !!table.options.enableColumnOrdering && columnOrderSettings && activeFilterValue.length === 0;
  const columnListItems = virtualItems.map((virtualItem) => {
    const column = filteredColumns[virtualItem.index];
    const columnKey = column.key;
    const canHide = getCanHide(table, column, isParentColumnHideable);
    const visibilityCheckboxState = getVisibilityCheckboxState(
      localColumnVisibility,
      column
    );
    const showColumnOrderDragHandle = canOrder && column.pinned === false;
    return /* @__PURE__ */ jsx(
      DataTableColumnsListItem,
      {
        column,
        columnIndex: localColumnOrder.indexOf(column.id),
        orderedFlatColumns,
        showVisibilityCheckbox,
        enableVisibilityCheckbox: canHide,
        currentVisibilityState: visibilityCheckboxState,
        onVisibilityChange: updateLocalColumnVisibility,
        showColumnOrderDragHandle,
        showColumnPinningMenu: showColumnPinningButton,
        localColumnPinning,
        updateLocalColumnPinning,
        virtualIndex: virtualItem.index,
        measureElement: columnVirtualizer.measureElement
      },
      columnKey
    );
  });
  const columnList = /* @__PURE__ */ jsx(
    "div",
    {
      ref: parentRef,
      className: clsx(
        "strato-table-settings-virtualization-container",
        styles.virtualizationScrollContainer({
          hasFilter: hasFilterBar
        })
      ),
      children: /* @__PURE__ */ jsx(
        Grid,
        {
          gridTemplateColumns: "1fr",
          padding: 0,
          gap: 0,
          className: clsx(styles.virtualizationInnerContainer),
          gridTemplateRows: columnVirtualizer.measurementsCache.map((el) => `${el.size}px`).join(" "),
          height: columnVirtualizer.getTotalSize() + 1,
          children: columnListItems
        }
      )
    }
  );
  return canOrder ? /* @__PURE__ */ jsx(
    ColumnOrderSettingsDragAndDropContext,
    {
      localColumnOrder,
      updateLocalColumnOrder,
      localColumnPinning,
      itemVirtualizer: columnVirtualizer,
      children: columnList
    }
  ) : columnList;
};
export {
  DataTableColumnsList
};
