// @dynatrace/strato-components/esm/tables/DataTable/components/Toolbar/ColumnSettings/DataTableColumnSettings.js
import { jsx, jsxs } from "react/jsx-runtime";
import { useMemo } from "react";
import { useIntl } from "react-intl";
import { ColumnsIcon } from "@dynatrace/strato-icons";
import { Button } from "../../../../../buttons/button/Button.js";
import { Tooltip } from "../../../../../overlays/tooltip/Tooltip.js";
import {
  useDataTable,
  useDataTableColumnSettings
} from "../../../contexts/DataTableContext.js";
import { dataTableMessages } from "../../../messages.js";
const DataTableColumnSettings = (props) => {
  const { customTriggerProps } = props;
  const {
    children: customTrigger,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    ...remainingProps
  } = customTriggerProps ?? {};
  const intl = useIntl();
  const table = useDataTable();
  const columnSettingsModal = useDataTableColumnSettings();
  const columnSettingsButton = useMemo(() => {
    const hiddenColumnCount = Object.values(
      table.getState().columnVisibility || {}
    ).filter((value) => !value).length;
    const message = hiddenColumnCount === 0 ? intl.formatMessage(dataTableMessages.columnSettingsButtonDefault) : intl.formatMessage(dataTableMessages.columnSettingsButtonVisibility, {
      hiddenColumnCount
    });
    return {
      text: message,
      hiddenColumnCount
    };
  }, [table.getState().columnVisibility]);
  return /* @__PURE__ */ jsx(Tooltip, { text: intl.formatMessage(dataTableMessages.columnSettingsTooltip), children: /* @__PURE__ */ jsxs(
    Button,
    {
      onClick: columnSettingsModal?.openColumnSettings,
      variant: "default",
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      ...remainingProps,
      children: [
        /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(ColumnsIcon, {}) }),
        customTrigger ? customTrigger({
          hiddenColumnCount: columnSettingsButton.hiddenColumnCount,
          totalColumnCount: table.getAllColumns().length,
          defaultLabel: columnSettingsButton.text
        }) : columnSettingsButton.text
      ]
    }
  ) });
};
export {
  DataTableColumnSettings
};
