// @dynatrace/strato-components/esm/tables/DataTable/components/Toolbar/ColumnSettings/DataTableColumnSettingsModal.js
import { jsx, jsxs } from "react/jsx-runtime";
import {
  forwardRef,
  useImperativeHandle,
  useMemo,
  useState
} from "react";
import { useIntl } from "react-intl";
import {
  generateColumnOrderWithParentIds,
  generateColumnVisibilityObject,
  getCombinedVisibilityCheckboxState,
  getHasColumnOrderChanged,
  getHasColumnPinningChanged,
  getHasVisibilityChanged,
  getMultipleLocalColumnVisibility,
  getResetOverride,
  getIsSettingEnabled
} from "./column-settings-utils.js";
import { DataTableColumnsList } from "./DataTableColumnsList.js";
import { DataTableVisibilityFilterControls } from "./DataTableVisibilityFilterControls.js";
import { useDataTable } from "../../../contexts/DataTableContext.js";
import { dataTableMessages } from "../../../messages.js";
import { DataTableSettingsModal } from "../../SettingsModal/DataTableSettingsModal.js";
const ColumnSettingsModal = (props) => {
  const {
    visibilitySettings: toolbarVisibilitySettings,
    columnOrderSettings: toolbarColumnOrderSettings,
    columnPinningSettings: toolbarColumnPinningSettings,
    resetColumnOrder,
    resetColumnVisibility,
    resetColumnPinning,
    onModalClose,
    columnSettingsOverride
  } = props;
  const intl = useIntl();
  const table = useDataTable();
  const leafColumnsWithoutBuiltin = table.getLeafColumnsWithoutBuiltin();
  const visibilityOverride = columnSettingsOverride?.columnVisibility;
  const columnOrderOverride = columnSettingsOverride?.columnOrder;
  const columnPinningOverride = columnSettingsOverride?.columnPinning;
  const visibilitySettingsEnabled = getIsSettingEnabled(
    visibilityOverride,
    toolbarVisibilitySettings
  );
  const columnOrderSettingsEnabled = getIsSettingEnabled(
    columnOrderOverride,
    toolbarColumnOrderSettings
  );
  const columnPinningSettingsEnabled = getIsSettingEnabled(
    columnPinningOverride,
    toolbarColumnPinningSettings
  );
  const {
    columnOrderWithParents: tableColumnOrder,
    uniqueParentIds: allLocalColumnsParentIds
  } = generateColumnOrderWithParentIds(
    table.getColumn,
    table.getState().columnOrder
  );
  const [activeFilterValue, setActiveFilterValue] = useState("");
  const initialColumnVisibility = useMemo(
    () => generateColumnVisibilityObject(leafColumnsWithoutBuiltin),
    [leafColumnsWithoutBuiltin]
  );
  const [localColumnVisibility, setLocalColumnVisibility] = useState(initialColumnVisibility);
  const [localColumnOrder, setLocalColumnOrder] = useState(tableColumnOrder);
  const [localColumnPinning, setLocalColumnPinning] = useState(
    // original column pinning state
    () => table.getState().columnPinning
  );
  const [showResetStatus, setShowResetStatus] = useState(false);
  const columnOrderToReset = getResetOverride(columnOrderOverride, "resetColumnOrder") ?? resetColumnOrder ?? table.options.originalColumnOrder;
  const columnVisibilityToReset = getResetOverride(visibilityOverride, "resetColumnVisibility") ?? resetColumnVisibility ?? table.options.originalColumnVisibility;
  const columnPinningToReset = getResetOverride(columnPinningOverride, "resetColumnPinning") ?? resetColumnPinning ?? table.options.originalColumnPinning;
  const filteredLocalColumnOrder = useMemo(
    () => localColumnOrder.filter(
      (colId) => !allLocalColumnsParentIds.includes(colId)
    ),
    [localColumnOrder, allLocalColumnsParentIds]
  );
  const hasVisibilityChanged = useMemo(
    () => visibilitySettingsEnabled && getHasVisibilityChanged(columnVisibilityToReset, localColumnVisibility),
    [visibilitySettingsEnabled, columnVisibilityToReset, localColumnVisibility]
  );
  const hasColumnOrderChanged = columnOrderSettingsEnabled && getHasColumnOrderChanged(columnOrderToReset, filteredLocalColumnOrder);
  const hasColumnPinningChanged = columnPinningSettingsEnabled && getHasColumnPinningChanged(columnPinningToReset, localColumnPinning);
  const showHideAllState = getCombinedVisibilityCheckboxState(
    localColumnVisibility,
    leafColumnsWithoutBuiltin
  );
  const updateLocalColumnVisibility = (column, currentVisibilityState) => {
    const childColumns = column.columns;
    if (childColumns.length === 0) {
      setLocalColumnVisibility((prevState) => ({
        ...prevState,
        [column.id]: !currentVisibilityState
      }));
    } else {
      setLocalColumnVisibility(
        (prevState) => getMultipleLocalColumnVisibility(
          childColumns,
          currentVisibilityState,
          prevState
        )
      );
    }
    showResetStatus && setShowResetStatus(false);
  };
  const updateLocalColumnOrder = (newColumnOrder) => {
    setLocalColumnOrder(newColumnOrder);
    showResetStatus && setShowResetStatus(false);
  };
  const updateLocalColumnPinning = (newState) => {
    setLocalColumnPinning(newState);
    showResetStatus && setShowResetStatus(false);
  };
  const onShowHideAllChange = () => {
    setLocalColumnVisibility(
      (prevState) => getMultipleLocalColumnVisibility(
        leafColumnsWithoutBuiltin,
        showHideAllState,
        prevState
      )
    );
    showResetStatus && setShowResetStatus(false);
  };
  const showVisibilityControl = visibilitySettingsEnabled && activeFilterValue.length === 0;
  const showFilterControl = visibilitySettingsEnabled || columnPinningSettingsEnabled;
  const onReset = () => {
    let didReset = false;
    if (visibilitySettingsEnabled && columnVisibilityToReset) {
      setLocalColumnVisibility(
        generateColumnVisibilityObject(
          leafColumnsWithoutBuiltin,
          columnVisibilityToReset
        )
      );
      didReset = true;
    }
    if (columnOrderSettingsEnabled && columnOrderToReset) {
      const { columnOrderWithParents } = generateColumnOrderWithParentIds(
        table.getColumn,
        columnOrderToReset
      );
      setLocalColumnOrder(columnOrderWithParents);
      didReset = true;
    }
    if (columnPinningSettingsEnabled && columnPinningToReset) {
      setLocalColumnPinning(columnPinningToReset);
      didReset = true;
    }
    didReset && setShowResetStatus(true);
  };
  const onApply = () => {
    table.setColumnVisibility(localColumnVisibility);
    table.setColumnOrder(filteredLocalColumnOrder);
    table.setColumnPinning(localColumnPinning);
    onModalClose();
  };
  return /* @__PURE__ */ jsxs(
    DataTableSettingsModal,
    {
      title: intl.formatMessage(dataTableMessages.columnSettingsTitle),
      show: true,
      showResetStatus,
      onDismiss: onModalClose,
      onReset,
      onApply,
      disableButtons: {
        reset: !hasColumnOrderChanged && !hasVisibilityChanged && !hasColumnPinningChanged
      },
      controlLabels: {
        reset: intl.formatMessage(
          dataTableMessages.resetColumnSettingsButtonText
        ),
        cancel: intl.formatMessage(dataTableMessages.cancelButtonText),
        apply: intl.formatMessage(dataTableMessages.applyButtonText)
      },
      children: [
        /* @__PURE__ */ jsx(
          DataTableVisibilityFilterControls,
          {
            showVisibilityControl,
            showFilterControl,
            setActiveFilterValue,
            showHideAllState,
            onShowHideAllChange
          }
        ),
        /* @__PURE__ */ jsx(
          DataTableColumnsList,
          {
            visibilitySettings: visibilitySettingsEnabled,
            columnOrderSettings: columnOrderSettingsEnabled,
            columnPinningSettings: columnPinningSettingsEnabled,
            activeFilterValue,
            localColumnVisibility,
            updateLocalColumnVisibility,
            localColumnOrder,
            updateLocalColumnOrder,
            localColumnPinning,
            updateLocalColumnPinning,
            hasFilterBar: showFilterControl || showVisibilityControl
          }
        )
      ]
    }
  );
};
function DataTableColumnSettingsModalFwd(props, forwardedRef) {
  const [modalOpen, setModalOpen] = useState(false);
  const [columnSettingsOverride, setColumnSettingsOverride] = useState();
  useImperativeHandle(forwardedRef, () => ({
    openColumnSettings: (columnSettings) => {
      setColumnSettingsOverride(columnSettings);
      setModalOpen(true);
    }
  }));
  if (modalOpen) {
    return /* @__PURE__ */ jsx(
      ColumnSettingsModal,
      {
        ...props,
        columnSettingsOverride,
        onModalClose: () => setModalOpen(false)
      }
    );
  }
  return null;
}
const DataTableColumnSettingsModal = forwardRef(
  DataTableColumnSettingsModalFwd
);
export {
  DataTableColumnSettingsModal
};
