// @dynatrace/strato-components/esm/charts/histogram/components/toolbar/DownloadDataButton.js
import { jsx } from "react/jsx-runtime";
import { useIntl } from "react-intl";
import { DownloadIcon } from "@dynatrace/strato-icons";
import { ChartToolbarRenderer } from "../../../core/components/toolbar/ChartToolbarRenderer.js";
import { useA11yToolbarTabIndex } from "../../../core/hooks/useA11yToolbarTabIndex.js";
import { useHistogramFilteredData } from "../../hooks/useHistogramFilteredData.js";
import { useHistogramToolbar } from "../../hooks/useHistogramToolbar.js";
import { HISTOGRAM_MESSAGES } from "../../messages/tooltip-text.js";
import { buildAndDownloadCsv } from "../../utils/build-and-download-csv.js";
const DownloadDataButton = () => {
  const intl = useIntl();
  const { downloadData } = useHistogramToolbar();
  const { series } = useHistogramFilteredData();
  const tabIndex = useA11yToolbarTabIndex("download");
  const handleKeyDown = (event) => {
    if (event.key === "Enter") {
      buildAndDownloadCsv(series);
    }
  };
  const handleClick = () => {
    buildAndDownloadCsv(series);
  };
  return downloadData?.enabled ? /* @__PURE__ */ jsx(
    ChartToolbarRenderer.Control,
    {
      prefixIcon: /* @__PURE__ */ jsx(DownloadIcon, {}),
      tooltipText: intl.formatMessage(HISTOGRAM_MESSAGES.downloadAsCsv),
      testId: "chart-toolbar-download",
      onKeyDown: handleKeyDown,
      onClick: handleClick,
      tabIndex,
      divider: "left"
    }
  ) : null;
};
DownloadDataButton["displayName"] = "DownloadDataButton";
export {
  DownloadDataButton
};
