// @dynatrace/strato-components/esm/charts/histogram/components/toolbar/HistogramToolbar.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { useIntl } from "react-intl";
import { DownloadDataButton } from "./DownloadDataButton.js";
import { ExploreButton } from "./ExploreButton.js";
import { PanButton } from "./PanButton.js";
import { ResetButton } from "./ResetButton.js";
import { ZoomInOutButtons } from "./ZoomInOutButtons.js";
import { ZoomModeButton } from "./ZoomModeButton.js";
import { useBehavioralTrackingProps as _useBehavioralTrackingProps } from "../../../../core/hooks/useBehavioralTrackingProps.js";
import { ChartToolbarRenderer } from "../../../core/components/toolbar/ChartToolbarRenderer.js";
import { getMenuTriggerDisplayMode } from "../../../core/components/toolbar/utils/get-menu-trigger-display-mode.js";
import { getMenuTriggerIcon } from "../../../core/components/toolbar/utils/get-menu-trigger-icon.js";
import { ChartOverlaysElevation } from "../../../core/constants.js";
import { getA11yToolbarConfig } from "../../../core/utils/zoom.js";
import { defaultToolbarPlacement } from "../../constants/defaults.js";
import { useInteractionModeState } from "../../hooks/store/selectors.js";
import { useHistogramFilteredData } from "../../hooks/useHistogramFilteredData.js";
import { useHistogramInteractions } from "../../hooks/useHistogramInteractions.js";
import { useHistogramSelectionAreaRef } from "../../hooks/useHistogramSelectionAreaRef.js";
import { useHistogramToolbar } from "../../hooks/useHistogramToolbar.js";
import { HISTOGRAM_MESSAGES } from "../../messages/tooltip-text.js";
import { buildAndDownloadCsv } from "../../utils/build-and-download-csv.js";
const HistogramToolbar = ({
  onMouseEnter,
  onMouseLeave,
  handlerRef
}) => {
  const toolbarContext = useHistogramToolbar();
  const { placement, collapsed, downloadData, intents } = toolbarContext;
  const interactionMode = useInteractionModeState();
  const { zoom, zoomX, pan } = useHistogramInteractions();
  const intl = useIntl();
  const selectionAreaRef = useHistogramSelectionAreaRef();
  const isZoomEnabled = zoom?.enabled;
  const isPanEnabled = pan?.enabled;
  const isDownloadEnabled = downloadData?.enabled ?? false;
  const anyInteractionEnabled = isZoomEnabled || zoomX?.enabled || isPanEnabled || isDownloadEnabled || intents;
  const menuIcon = getMenuTriggerIcon(interactionMode, isZoomEnabled);
  const menuLabel = !isPanEnabled && !isZoomEnabled ? intl.formatMessage(HISTOGRAM_MESSAGES.zoomModeExplore) : void 0;
  const menuDisplayMode = getMenuTriggerDisplayMode({
    isPanEnabled,
    isZoomEnabled,
    isDownloadEnabled
  });
  const onClose = () => selectionAreaRef ? selectionAreaRef.current?.focus() : void 0;
  const { series } = useHistogramFilteredData();
  const handleDownloadDataToolbar = () => buildAndDownloadCsv(series);
  const showDefaultToolbarDownloadButton = isDownloadEnabled && !intents;
  const showMenuToolbarButton = isDownloadEnabled && !!intents?.length;
  const behaviorTrackingProps = _useBehavioralTrackingProps(toolbarContext)[0];
  return anyInteractionEnabled ? /* @__PURE__ */ jsxs(
    ChartToolbarRenderer,
    {
      visible: false,
      visibleWhenMouseInBounds: true,
      handlerRef,
      placement: placement || defaultToolbarPlacement,
      collapsed,
      draggable: true,
      initialA11yState: getA11yToolbarConfig({
        zoomX,
        pan,
        zoom,
        download: downloadData,
        intent: !!intents
      }),
      zIndex: ChartOverlaysElevation.toolbar,
      onMouseEnter,
      onMouseLeave,
      menuIcon,
      menuTriggerLabel: menuLabel,
      menuDisplayMode,
      onClose,
      isDownloadEnabled: showMenuToolbarButton,
      intents,
      handleDownloadDataToolbar,
      ...behaviorTrackingProps,
      children: [
        /* @__PURE__ */ jsxs(Fragment, { children: [
          /* @__PURE__ */ jsx(ExploreButton, {}),
          isZoomEnabled && /* @__PURE__ */ jsx(ZoomModeButton, {}),
          isPanEnabled && /* @__PURE__ */ jsx(PanButton, {})
        ] }),
        isZoomEnabled && /* @__PURE__ */ jsx(ZoomInOutButtons, {}),
        /* @__PURE__ */ jsx(ResetButton, {}),
        showDefaultToolbarDownloadButton && /* @__PURE__ */ jsx(DownloadDataButton, {})
      ]
    }
  ) : null;
};
export {
  HistogramToolbar
};
