// @dynatrace/strato-components/esm/charts/histogram/components/toolbar/ResetButton.js
import { jsx } from "react/jsx-runtime";
import { useMemo } from "react";
import { useIntl } from "react-intl";
import { ResetIcon } from "@dynatrace/strato-icons";
import { reduceExploreCancelSelect } from "../../../core/components/interactions-state-helpers/reducers/explore-reducers.js";
import { ChartToolbarRenderer } from "../../../core/components/toolbar/ChartToolbarRenderer.js";
import { useA11yToolbarTabIndex } from "../../../core/hooks/useA11yToolbarTabIndex.js";
import { isSameDomain } from "../../../core/utils/zoom.js";
import { useCurrentDomainState } from "../../hooks/store/selectors.js";
import { useSetState } from "../../hooks/store/state-setters.js";
import { useHistogramInteractions } from "../../hooks/useHistogramInteractions.js";
import { useHistogramXScale } from "../../hooks/useHistogramXScale.js";
import { useResetInteractions } from "../../hooks/useResetInteractions.js";
import { HISTOGRAM_MESSAGES } from "../../messages/tooltip-text.js";
const ResetButton = () => {
  const tabIndex = useA11yToolbarTabIndex("reset");
  const intl = useIntl();
  const { initialXScale } = useHistogramXScale();
  const setState = useSetState();
  const currentDomain = useCurrentDomainState();
  const { zoom, zoomX } = useHistogramInteractions();
  const hasZoomInteractions = zoom?.enabled || zoomX?.enabled;
  const resetInteractions = useResetInteractions();
  const handleKeyDown = (event) => {
    if (event.code === "Enter") {
      resetInteractions();
    }
  };
  const handleClick = () => {
    setState(reduceExploreCancelSelect);
    resetInteractions();
  };
  const isResettable = useMemo(
    () => !isSameDomain(currentDomain, initialXScale.domain()),
    [currentDomain, initialXScale]
  );
  const onlyXZoomIsEnabled = zoomX?.enabled && !zoom?.enabled;
  return hasZoomInteractions ? /* @__PURE__ */ jsx(
    ChartToolbarRenderer.Control,
    {
      prefixIcon: /* @__PURE__ */ jsx(ResetIcon, {}),
      tooltipText: intl.formatMessage(HISTOGRAM_MESSAGES.resetMode),
      tooltipKeyShortcut: "R",
      onKeyDown: handleKeyDown,
      onClick: handleClick,
      tabIndex,
      disabled: !isResettable,
      placement: "toolbar",
      divider: onlyXZoomIsEnabled ? "left" : void 0,
      showWhenCollapsed: isResettable
    }
  ) : null;
};
ResetButton["displayName"] = "ResetButton";
export {
  ResetButton
};
