// @dynatrace/strato-components/esm/charts/core/components/single-value-renderer/components/Label.js
import { jsx } from "react/jsx-runtime";
import { useContext, useLayoutEffect, useRef } from "react";
import { labelAreaStaticStyles } from "./Label.sty.js";
import { useId } from "../../../../../core/hooks/useId.js";
import { useMergeRefs } from "../../../../../core/hooks/useMergeRefs.js";
import { Flex } from "../../../../../layouts/flex/Flex.js";
import { useFontSizeForContainer } from "../../../hooks/useFontSizeForContainer.js";
import { useThrottledResizeObserver } from "../../../hooks/useThrottledResizeObserver.js";
import { FONT_HEIGHT_RATIO, LABEL_FONT_CSS_TOKEN } from "../constants.js";
import { SingleValueContext } from "../contexts/single-value.context.js";
import { useGlobalFontSize } from "../hooks/use-global-font-size.js";
const Label = ({ color }) => {
  const { label, alignment } = useContext(SingleValueContext);
  const ref = useRef(null);
  const { refCallback, width = 0, height = 0 } = useThrottledResizeObserver();
  const mergedRef = useMergeRefs([refCallback, ref]);
  const preferredFontSize = useFontSizeForContainer({
    container: { width, height: height * FONT_HEIGHT_RATIO },
    text: label ?? "",
    element: ref.current
  });
  const id = useId();
  const { updateLabelFontSize } = useGlobalFontSize();
  useLayoutEffect(() => {
    if (height && width && label) {
      updateLabelFontSize?.(id, preferredFontSize);
    }
  }, [height, width, label, preferredFontSize, updateLabelFontSize, id]);
  return /* @__PURE__ */ jsx(
    Flex,
    {
      className: labelAreaStaticStyles,
      style: {
        [LABEL_FONT_CSS_TOKEN]: preferredFontSize,
        color
      },
      ref: mergedRef,
      justifyContent: alignment,
      alignItems: "center",
      children: label !== void 0 && /* @__PURE__ */ jsx("div", { "data-testid": "single-value-chart-label", children: label })
    }
  );
};
Label["displayName"] = "SingleValue.Label";
export {
  Label
};
