// @dynatrace/strato-components/esm/charts/core/components/single-value-renderer/components/SingleValueToolbar.js
import { jsx } from "react/jsx-runtime";
import { toolbarOverlayPadding } from "./SingleValueToolbar.sty.js";
import { ChartOverlaysElevation } from "../../../constants.js";
import { getA11yToolbarConfig } from "../../../utils/zoom.js";
import { ChartToolbarRenderer } from "../../toolbar/ChartToolbarRenderer.js";
import {
  MINIMUM_HEIGHT_FOR_TOOLBAR,
  MINIMUM_HEIGHT_FOR_TOOLBAR_WITH_PADDING
} from "../constants.js";
const SingleValueToolbar = (props) => {
  const {
    parentHeight,
    hidden,
    intents,
    intentsAsyncTemplate,
    placement = "top-right",
    draggable
  } = props;
  const hasEnoughHeightForToolbar = parentHeight > MINIMUM_HEIGHT_FOR_TOOLBAR;
  const isToolbarEnabled = hasEnoughHeightForToolbar && !hidden && !!(intents?.length ?? intentsAsyncTemplate);
  if (!isToolbarEnabled) {
    return null;
  }
  const overlayClassName = parentHeight > MINIMUM_HEIGHT_FOR_TOOLBAR_WITH_PADDING ? void 0 : toolbarOverlayPadding;
  return /* @__PURE__ */ jsx(
    ChartToolbarRenderer,
    {
      visible: false,
      placement,
      initialA11yState: getA11yToolbarConfig({
        intent: !!intents
      }),
      menuDisplayMode: "never",
      intents,
      overlayClassName,
      visibleWhenMouseInBounds: true,
      intentsAsyncTemplate,
      zIndex: ChartOverlaysElevation.toolbar,
      draggable,
      removeDragHandle: true
    }
  );
};
SingleValueToolbar["displayName"] = "SingleValueToolbar";
export {
  SingleValueToolbar
};
