// @dynatrace/strato-components/esm/charts/gauge/providers/GaugeChartProviders.js
import { jsx } from "react/jsx-runtime";
import { useMemo } from "react";
import { OverlayProvider } from "../../core/components/overlay-tooltip/contexts/OverlayContext.js";
import { TooltipProvider } from "../../core/components/overlay-tooltip/contexts/TooltipContext.js";
import { ChartSeriesActionContext } from "../../core/components/series-action/ChartSeriesActionContext.js";
import { GaugeChartColorContext } from "../contexts/GaugeChartColorContext.js";
import { GaugeChartFormatterContext } from "../contexts/GaugeChartFormatterContext.js";
import { GaugeChartLabelsContext } from "../contexts/GaugeChartLabelsContext.js";
import { GaugeChartLoadingContext } from "../contexts/GaugeChartLoadingContext.js";
import { GaugeChartRangeMaxContext } from "../contexts/GaugeChartRangeMaxContext.js";
import { GaugeChartRangeMinContext } from "../contexts/GaugeChartRangeMinContext.js";
import { GaugeChartSizeContext } from "../contexts/GaugeChartSizeContext.js";
import { GaugeChartThresholdContext } from "../contexts/GaugeChartThresholdContext.js";
import { GaugeChartTooltipContext } from "../contexts/GaugeChartTooltipContext.js";
import { getValueArcColor } from "../utils/get-value-arc-color.js";
const emptyThresholds = [];
const GaugeChartProviders = ({
  config,
  width,
  height,
  value,
  valueAccessor,
  explicitColor,
  seriesActions,
  children,
  loading
}) => {
  const {
    formatter,
    max,
    min,
    labelSlot,
    thresholdsIndicators,
    colorRules,
    unit
  } = config;
  const calculatedColor = getValueArcColor(
    value,
    colorRules,
    explicitColor,
    thresholdsIndicators,
    valueAccessor
  );
  const seriesActionValue = useMemo(
    () => ({
      chartSeriesActions: seriesActions
    }),
    [seriesActions]
  );
  const chartSizeValue = useMemo(() => ({ width, height }), [width, height]);
  const formatterContextValue = useMemo(
    () => ({
      formatter,
      unit
    }),
    [formatter, unit]
  );
  const tooltipConfigValue = useMemo(
    () => ({
      name: config.tooltip?.name,
      template: config.tooltip?.children,
      hidden: config.tooltip?.hidden ?? false
    }),
    [config.tooltip]
  );
  return /* @__PURE__ */ jsx(OverlayProvider, { children: /* @__PURE__ */ jsx(TooltipProvider, { persistedStore: true, children: /* @__PURE__ */ jsx(ChartSeriesActionContext.Provider, { value: seriesActionValue, children: /* @__PURE__ */ jsx(GaugeChartSizeContext.Provider, { value: chartSizeValue, children: /* @__PURE__ */ jsx(GaugeChartFormatterContext.Provider, { value: formatterContextValue, children: /* @__PURE__ */ jsx(
    GaugeChartThresholdContext.Provider,
    {
      value: thresholdsIndicators ?? emptyThresholds,
      children: /* @__PURE__ */ jsx(GaugeChartColorContext.Provider, { value: calculatedColor, children: /* @__PURE__ */ jsx(
        GaugeChartLabelsContext.Provider,
        {
          value: {
            label: labelSlot
          },
          children: /* @__PURE__ */ jsx(GaugeChartRangeMinContext.Provider, { value: min, children: /* @__PURE__ */ jsx(GaugeChartRangeMaxContext.Provider, { value: max, children: /* @__PURE__ */ jsx(GaugeChartLoadingContext.Provider, { value: loading, children: /* @__PURE__ */ jsx(
            GaugeChartTooltipContext.Provider,
            {
              value: tooltipConfigValue,
              children
            }
          ) }) }) })
        }
      ) })
    }
  ) }) }) }) }) });
};
GaugeChartProviders["displayName"] = "GaugeChartProviders";
export {
  GaugeChartProviders
};
