// @dynatrace/strato-components/esm/forms/text-input/TextInput.js
import { jsx } from "react/jsx-runtime";
import { forwardRef } from "react";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useControllableState } from "../../core/hooks/useControllableState.js";
import { mergeProps } from "../../core/utils/merge-props.js";
import { BaseInput } from "../_base-input/BaseInput.js";
import { BaseInputButton } from "../_base-input/BaseInputButton.js";
import {
  BaseInputSuffix,
  BaseInputPrefix
} from "../_base-input/BaseInputContent.js";
const COMPONENT_NAME = "TextInput";
const TextInputRoot = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const {
    onChange,
    value,
    defaultValue,
    required,
    readOnly,
    type = "text",
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "aria-disabled": ariaDisabledProp,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
  const [internalValue, setInternalValue] = useControllableState({
    defaultValue,
    value
  });
  const handleOnChange = (e) => {
    onChange?.(e.target.value, e);
    setInternalValue(e.target.value);
  };
  return /* @__PURE__ */ jsx(
    BaseInput,
    {
      ref: forwardedRef,
      ...mergeProps(restProps, behavioralTrackingProps),
      type,
      value: internalValue,
      required,
      readOnly,
      onChange: handleOnChange,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "aria-disabled": ariaDisabledProp,
      "data-dt-component": COMPONENT_NAME
    }
  );
});
const TextInput = Object.assign(TextInputRoot, {
  Prefix: BaseInputPrefix,
  Suffix: BaseInputSuffix,
  Button: BaseInputButton
});
TextInput.displayName = COMPONENT_NAME;
export {
  TextInput
};
