// @dynatrace/strato-components/esm/tables/DataTable/tanstack-table-core-override/src/features/ColumnPinning.js
import { getRowProto } from "..";
import { getMemoOptions, makeStateUpdater, memo } from "../utils";
const getDefaultColumnPinningState = () => ({
  left: [],
  right: []
});
const ColumnPinning = {
  getInitialState: (state) => {
    return {
      columnPinning: getDefaultColumnPinningState(),
      ...state
    };
  },
  getDefaultOptions: (table) => {
    return {
      onColumnPinningChange: makeStateUpdater("columnPinning", table)
    };
  },
  createColumn: (column, table) => {
    column.pin = (position) => {
      const columnIds = column.getLeafColumns().map((d) => d.id).filter(Boolean);
      table.setColumnPinning((old) => {
        if (position === "right") {
          return {
            left: (old?.left ?? []).filter((d) => !columnIds?.includes(d)),
            right: [
              ...(old?.right ?? []).filter((d) => !columnIds?.includes(d)),
              ...columnIds
            ]
          };
        }
        if (position === "left") {
          return {
            left: [
              ...(old?.left ?? []).filter((d) => !columnIds?.includes(d)),
              ...columnIds
            ],
            right: (old?.right ?? []).filter((d) => !columnIds?.includes(d))
          };
        }
        return {
          left: (old?.left ?? []).filter((d) => !columnIds?.includes(d)),
          right: (old?.right ?? []).filter((d) => !columnIds?.includes(d))
        };
      });
    };
    column.getCanPin = () => {
      const leafColumns = column.getLeafColumns();
      return leafColumns.some(
        (d) => (d.columnDef.enablePinning ?? true) && (table.options.enableColumnPinning ?? table.options.enablePinning ?? true)
      );
    };
    column.getIsPinned = () => {
      const leafColumnIds = column.getLeafColumns().map((d) => d.id);
      const { left, right } = table.getState().columnPinning;
      const isLeft = leafColumnIds.some((d) => left?.includes(d));
      const isRight = leafColumnIds.some((d) => right?.includes(d));
      return isLeft ? "left" : isRight ? "right" : false;
    };
    column.getPinnedIndex = () => {
      const position = column.getIsPinned();
      return position ? table.getState().columnPinning?.[position]?.indexOf(column.id) ?? -1 : 0;
    };
  },
  createTable: (table) => {
    table.setColumnPinning = (updater) => table.options.onColumnPinningChange?.(updater);
    table.resetColumnPinning = (defaultState) => table.setColumnPinning(
      defaultState ? getDefaultColumnPinningState() : table.initialState?.columnPinning ?? getDefaultColumnPinningState()
    );
    table.getIsSomeColumnsPinned = (position) => {
      const pinningState = table.getState().columnPinning;
      if (!position) {
        return Boolean(pinningState.left?.length || pinningState.right?.length);
      }
      return Boolean(pinningState[position]?.length);
    };
    table.getLeftLeafColumns = memo(
      () => [table.getAllLeafColumns(), table.getState().columnPinning.left],
      (allColumns, left) => {
        return (left ?? []).map(
          (columnId) => allColumns.find((column) => column.id === columnId)
        ).filter(Boolean);
      },
      getMemoOptions(table.options, "debugColumns", "getLeftLeafColumns")
    );
    table.getRightLeafColumns = memo(
      () => [table.getAllLeafColumns(), table.getState().columnPinning.right],
      (allColumns, right) => {
        return (right ?? []).map(
          (columnId) => allColumns.find((column) => column.id === columnId)
        ).filter(Boolean);
      },
      getMemoOptions(table.options, "debugColumns", "getRightLeafColumns")
    );
    table.getCenterLeafColumns = memo(
      () => [
        table.getAllLeafColumns(),
        table.getState().columnPinning.left,
        table.getState().columnPinning.right
      ],
      (allColumns, left, right) => {
        const leftAndRight = [...left ?? [], ...right ?? []];
        return allColumns.filter((d) => !leftAndRight.includes(d.id));
      },
      getMemoOptions(table.options, "debugColumns", "getCenterLeafColumns")
    );
    Object.assign(getRowProto(table), {
      getCenterVisibleCells: memo(
        function() {
          return [
            this._getAllVisibleCells(),
            table.getState().columnPinning.left,
            table.getState().columnPinning.right
          ];
        },
        (allCells, left, right) => {
          const leftAndRight = [...left ?? [], ...right ?? []];
          return allCells.filter((d) => !leftAndRight.includes(d.column.id));
        },
        getMemoOptions(table.options, "debugRows", "getCenterVisibleCells")
      ),
      getLeftVisibleCells: memo(
        function() {
          return [
            this._getAllVisibleCells(),
            table.getState().columnPinning.left
          ];
        },
        (allCells, left) => {
          const cells = (left ?? []).map(
            (columnId) => allCells.find((cell) => cell.column.id === columnId)
          ).filter(Boolean).map((d) => ({ ...d, position: "left" }));
          return cells;
        },
        getMemoOptions(table.options, "debugRows", "getLeftVisibleCells")
      ),
      getRightVisibleCells: memo(
        function() {
          return [
            this._getAllVisibleCells(),
            table.getState().columnPinning.right
          ];
        },
        (allCells, right) => {
          const cells = (right ?? []).map(
            (columnId) => allCells.find((cell) => cell.column.id === columnId)
          ).filter(Boolean).map((d) => ({ ...d, position: "right" }));
          return cells;
        },
        getMemoOptions(table.options, "debugRows", "getRightVisibleCells")
      )
    });
  }
};
export {
  ColumnPinning
};
