// @dynatrace/strato-components/esm/tables/DataTable/tanstack-table-core-override/src/features/ColumnFiltering.js
import { getRowProto } from "..";
import { filterFns } from "../filterFns";
import { functionalUpdate, isFunction, makeStateUpdater } from "../utils";
const ColumnFiltering = {
  getDefaultColumnDef: () => {
    return {
      filterFn: "auto"
    };
  },
  getInitialState: (state) => {
    return {
      columnFilters: [],
      ...state
    };
  },
  getDefaultOptions: (table) => {
    return {
      onColumnFiltersChange: makeStateUpdater("columnFilters", table),
      filterFromLeafRows: false,
      maxLeafRowFilterDepth: 100
    };
  },
  createColumn: (column, table) => {
    column.getAutoFilterFn = () => {
      const firstRow = table.getCoreRowModel().flatRows[0];
      const value = firstRow?.getValue(column.id);
      if (typeof value === "string") {
        return filterFns.includesString;
      }
      if (typeof value === "number") {
        return filterFns.inNumberRange;
      }
      if (typeof value === "boolean") {
        return filterFns.equals;
      }
      if (value !== null && typeof value === "object") {
        return filterFns.equals;
      }
      if (Array.isArray(value)) {
        return filterFns.arrIncludes;
      }
      return filterFns.weakEquals;
    };
    column.getFilterFn = () => {
      return isFunction(column.columnDef.filterFn) ? column.columnDef.filterFn : column.columnDef.filterFn === "auto" ? column.getAutoFilterFn() : (
        // @ts-ignore
        table.options.filterFns?.[column.columnDef.filterFn] ?? filterFns[column.columnDef.filterFn]
      );
    };
    column.getCanFilter = () => {
      return (column.columnDef.enableColumnFilter ?? true) && (table.options.enableColumnFilters ?? true) && (table.options.enableFilters ?? true) && !!column.accessorFn;
    };
    column.getIsFiltered = () => column.getFilterIndex() > -1;
    column.getFilterValue = () => table.getState().columnFilters?.find((d) => d.id === column.id)?.value;
    column.getFilterIndex = () => table.getState().columnFilters?.findIndex((d) => d.id === column.id) ?? -1;
    column.setFilterValue = (value) => {
      table.setColumnFilters((old) => {
        const filterFn = column.getFilterFn();
        const previousFilter = old?.find((d) => d.id === column.id);
        const newFilter = functionalUpdate(
          value,
          previousFilter ? previousFilter.value : void 0
        );
        if (shouldAutoRemoveFilter(filterFn, newFilter, column)) {
          return old?.filter((d) => d.id !== column.id) ?? [];
        }
        const newFilterObj = { id: column.id, value: newFilter };
        if (previousFilter) {
          return old?.map((d) => {
            if (d.id === column.id) {
              return newFilterObj;
            }
            return d;
          }) ?? [];
        }
        if (old?.length) {
          return [...old, newFilterObj];
        }
        return [newFilterObj];
      });
    };
  },
  createTable: (table) => {
    table.setColumnFilters = (updater) => {
      const leafColumns = table.getAllLeafColumns();
      const updateFn = (old) => {
        return functionalUpdate(updater, old)?.filter((filter) => {
          const column = leafColumns.find((d) => d.id === filter.id);
          if (column) {
            const filterFn = column.getFilterFn();
            if (shouldAutoRemoveFilter(filterFn, filter.value, column)) {
              return false;
            }
          }
          return true;
        });
      };
      table.options.onColumnFiltersChange?.(updateFn);
    };
    table.resetColumnFilters = (defaultState) => {
      table.setColumnFilters(
        defaultState ? [] : table.initialState?.columnFilters ?? []
      );
    };
    table.getPreFilteredRowModel = () => table.getCoreRowModel();
    table.getFilteredRowModel = () => {
      if (!table._getFilteredRowModel && table.options.getFilteredRowModel) {
        table._getFilteredRowModel = table.options.getFilteredRowModel(table);
      }
      if (table.options.manualFiltering || !table._getFilteredRowModel) {
        return table.getPreFilteredRowModel();
      }
      return table._getFilteredRowModel();
    };
    Object.defineProperties(getRowProto(table), {
      columnFilters: {
        get() {
          return this._columnFilters ??= {};
        },
        set(value) {
          this._columnFilters = value;
        },
        enumerable: true
      },
      columnFiltersMeta: {
        get() {
          return this._columnFiltersMeta ??= {};
        },
        set(value) {
          this._columnFiltersMeta = value;
        },
        enumerable: true
      }
    });
  }
};
function shouldAutoRemoveFilter(filterFn, value, column) {
  return (filterFn && filterFn.autoRemove ? filterFn.autoRemove(value, column) : false) || typeof value === "undefined" || typeof value === "string" && !value;
}
export {
  ColumnFiltering,
  shouldAutoRemoveFilter
};
