// @dynatrace/strato-components/esm/tables/DataTable/tanstack-table-core-override/src/features/ColumnGrouping.js
import { getRowProto } from "..";
import { aggregationFns } from "../aggregationFns";
import { isFunction, makeStateUpdater } from "../utils";
const ColumnGrouping = {
  getDefaultColumnDef: () => {
    return {
      aggregatedCell: (props) => props.getValue()?.toString?.() ?? null,
      aggregationFn: "auto"
    };
  },
  getInitialState: (state) => {
    return {
      grouping: [],
      ...state
    };
  },
  getDefaultOptions: (table) => {
    return {
      onGroupingChange: makeStateUpdater("grouping", table),
      groupedColumnMode: "reorder"
    };
  },
  createColumn: (column, table) => {
    column.toggleGrouping = () => {
      table.setGrouping((old) => {
        if (old?.includes(column.id)) {
          return old.filter((d) => d !== column.id);
        }
        return [...old ?? [], column.id];
      });
    };
    column.getCanGroup = () => {
      return (column.columnDef.enableGrouping ?? true) && (table.options.enableGrouping ?? true) && (!!column.accessorFn || !!column.columnDef.getGroupingValue);
    };
    column.getIsGrouped = () => {
      return table.getState().grouping?.includes(column.id);
    };
    column.getGroupedIndex = () => table.getState().grouping?.indexOf(column.id);
    column.getToggleGroupingHandler = () => {
      const canGroup = column.getCanGroup();
      return () => {
        if (!canGroup) return;
        column.toggleGrouping();
      };
    };
    column.getAutoAggregationFn = () => {
      const firstRow = table.getCoreRowModel().flatRows[0];
      const value = firstRow?.getValue(column.id);
      if (typeof value === "number") {
        return aggregationFns.sum;
      }
      if (Object.prototype.toString.call(value) === "[object Date]") {
        return aggregationFns.extent;
      }
    };
    column.getAggregationFn = () => {
      if (!column) {
        throw new Error();
      }
      return isFunction(column.columnDef.aggregationFn) ? column.columnDef.aggregationFn : column.columnDef.aggregationFn === "auto" ? column.getAutoAggregationFn() : table.options.aggregationFns?.[column.columnDef.aggregationFn] ?? aggregationFns[column.columnDef.aggregationFn];
    };
  },
  createTable: (table) => {
    table.setGrouping = (updater) => table.options.onGroupingChange?.(updater);
    table.resetGrouping = (defaultState) => {
      table.setGrouping(defaultState ? [] : table.initialState?.grouping ?? []);
    };
    table.getPreGroupedRowModel = () => table.getFilteredRowModel();
    table.getGroupedRowModel = () => {
      if (!table._getGroupedRowModel && table.options.getGroupedRowModel) {
        table._getGroupedRowModel = table.options.getGroupedRowModel(table);
      }
      if (table.options.manualGrouping || !table._getGroupedRowModel) {
        return table.getPreGroupedRowModel();
      }
      return table._getGroupedRowModel();
    };
    Object.defineProperty(getRowProto(table), "_groupingValuesCache", {
      get() {
        return this.__groupingValuesCache ??= {};
      },
      enumerable: true
    });
    Object.assign(getRowProto(table), {
      getIsGrouped() {
        return !!this.groupingColumnId;
      },
      getGroupingValue(columnId) {
        if (this._groupingValuesCache.hasOwnProperty(columnId)) {
          return this._groupingValuesCache[columnId];
        }
        const column = table.getColumn(columnId);
        if (!column?.columnDef.getGroupingValue) {
          return this.getValue(columnId);
        }
        this._groupingValuesCache[columnId] = column.columnDef.getGroupingValue(
          this.original
        );
        return this._groupingValuesCache[columnId];
      }
    });
  },
  createCell: (cell, column, row, table) => {
    const getRenderValue = () => cell.getValue() ?? table.options.renderFallbackValue;
    cell.getIsGrouped = () => column.getIsGrouped() && column.id === row.groupingColumnId;
    cell.getIsPlaceholder = () => !cell.getIsGrouped() && column.getIsGrouped();
    cell.getIsAggregated = () => !cell.getIsGrouped() && !cell.getIsPlaceholder() && !!row.subRows?.length;
  }
};
function orderColumns(leafColumns, grouping, groupedColumnMode) {
  if (!grouping?.length || !groupedColumnMode) {
    return leafColumns;
  }
  const nonGroupingColumns = leafColumns.filter(
    (col) => !grouping.includes(col.id)
  );
  if (groupedColumnMode === "remove") {
    return nonGroupingColumns;
  }
  const groupingColumns = grouping.map((g) => leafColumns.find((col) => col.id === g)).filter(Boolean);
  return [...groupingColumns, ...nonGroupingColumns];
}
export {
  ColumnGrouping,
  orderColumns
};
