// @dynatrace/strato-components/esm/tables/DataTable/tanstack-table-core-override/src/features/ColumnVisibility.js
import { getRowProto } from "..";
import { getMemoOptions, makeStateUpdater, memo } from "../utils";
const ColumnVisibility = {
  getInitialState: (state) => {
    return {
      columnVisibility: {},
      ...state
    };
  },
  getDefaultOptions: (table) => {
    return {
      onColumnVisibilityChange: makeStateUpdater("columnVisibility", table)
    };
  },
  createColumn: (column, table) => {
    column.toggleVisibility = (value) => {
      if (column.getCanHide()) {
        table.setColumnVisibility((old) => ({
          ...old,
          [column.id]: value ?? !column.getIsVisible()
        }));
      }
    };
    column.getIsVisible = () => {
      const childColumns = column.columns;
      return (childColumns.length ? childColumns.some((c) => c.getIsVisible()) : table.getState().columnVisibility?.[column.id]) ?? true;
    };
    column.getCanHide = () => {
      return (column.columnDef.enableHiding ?? true) && (table.options.enableHiding ?? true);
    };
    column.getToggleVisibilityHandler = () => {
      return (e) => {
        column.toggleVisibility?.(
          e.target.checked
        );
      };
    };
  },
  createTable: (table) => {
    const makeVisibleColumnsMethod = (key, getColumns) => {
      return memo(
        () => [
          getColumns(),
          getColumns().filter((d) => d.getIsVisible()).map((d) => d.id).join("_")
        ],
        (columns) => {
          return columns.filter((d) => d.getIsVisible?.());
        },
        getMemoOptions(table.options, "debugColumns", key)
      );
    };
    table.getVisibleFlatColumns = makeVisibleColumnsMethod(
      "getVisibleFlatColumns",
      () => table.getAllFlatColumns()
    );
    table.getVisibleLeafColumns = makeVisibleColumnsMethod(
      "getVisibleLeafColumns",
      () => table.getAllLeafColumns()
    );
    table.getLeftVisibleLeafColumns = makeVisibleColumnsMethod(
      "getLeftVisibleLeafColumns",
      () => table.getLeftLeafColumns()
    );
    table.getRightVisibleLeafColumns = makeVisibleColumnsMethod(
      "getRightVisibleLeafColumns",
      () => table.getRightLeafColumns()
    );
    table.getCenterVisibleLeafColumns = makeVisibleColumnsMethod(
      "getCenterVisibleLeafColumns",
      () => table.getCenterLeafColumns()
    );
    table.setColumnVisibility = (updater) => table.options.onColumnVisibilityChange?.(updater);
    table.resetColumnVisibility = (defaultState) => {
      table.setColumnVisibility(
        defaultState ? {} : table.initialState.columnVisibility ?? {}
      );
    };
    table.toggleAllColumnsVisible = (value) => {
      value = value ?? !table.getIsAllColumnsVisible();
      table.setColumnVisibility(
        table.getAllLeafColumns().reduce(
          (obj, column) => ({
            ...obj,
            [column.id]: !value ? !column.getCanHide?.() : value
          }),
          {}
        )
      );
    };
    table.getIsAllColumnsVisible = () => !table.getAllLeafColumns().some((column) => !column.getIsVisible?.());
    table.getIsSomeColumnsVisible = () => table.getAllLeafColumns().some((column) => column.getIsVisible?.());
    table.getToggleAllColumnsVisibilityHandler = () => {
      return (e) => {
        table.toggleAllColumnsVisible(
          e.target?.checked
        );
      };
    };
    Object.assign(getRowProto(table), {
      _getAllVisibleCells: memo(
        function() {
          return [this.getAllCells(), table.getState().columnVisibility];
        },
        (cells) => {
          return cells.filter((cell) => cell.column.getIsVisible());
        },
        getMemoOptions(table.options, "debugRows", "_getAllVisibleCells")
      ),
      getVisibleCells: memo(
        function() {
          return [
            this.getLeftVisibleCells(),
            this.getCenterVisibleCells(),
            this.getRightVisibleCells()
          ];
        },
        (left, center, right) => [...left, ...center, ...right],
        getMemoOptions(table.options, "debugRows", "getVisibleCells")
      )
    });
  }
};
function _getVisibleLeafColumns(table, position) {
  return !position ? table.getVisibleLeafColumns() : position === "center" ? table.getCenterVisibleLeafColumns() : position === "left" ? table.getLeftVisibleLeafColumns() : table.getRightVisibleLeafColumns();
}
export {
  ColumnVisibility,
  _getVisibleLeafColumns
};
