// @dynatrace/strato-components/esm/tables/DataTable/tanstack-table-core-override/src/features/ColumnOrdering.js
import { getMemoOptions, makeStateUpdater, memo } from "../utils";
import { orderColumns } from "./ColumnGrouping";
import { _getVisibleLeafColumns } from "..";
const ColumnOrdering = {
  getInitialState: (state) => {
    return {
      columnOrder: [],
      ...state
    };
  },
  getDefaultOptions: (table) => {
    return {
      onColumnOrderChange: makeStateUpdater("columnOrder", table)
    };
  },
  createColumn: (column, table) => {
    column.getIndex = memo(
      (position) => [_getVisibleLeafColumns(table, position)],
      (columns) => columns.findIndex((d) => d.id === column.id),
      getMemoOptions(table.options, "debugColumns", "getIndex")
    );
    column.getIsFirstColumn = (position) => {
      const columns = _getVisibleLeafColumns(table, position);
      return columns[0]?.id === column.id;
    };
    column.getIsLastColumn = (position) => {
      const columns = _getVisibleLeafColumns(table, position);
      return columns[columns.length - 1]?.id === column.id;
    };
  },
  createTable: (table) => {
    table.setColumnOrder = (updater) => table.options.onColumnOrderChange?.(updater);
    table.resetColumnOrder = (defaultState) => {
      table.setColumnOrder(
        defaultState ? [] : table.initialState.columnOrder ?? []
      );
    };
    table._getOrderColumnsFn = memo(
      () => [
        table.getState().columnOrder,
        table.getState().grouping,
        table.options.groupedColumnMode
      ],
      (columnOrder, grouping, groupedColumnMode) => (columns) => {
        let orderedColumns = [];
        if (!columnOrder?.length) {
          orderedColumns = columns;
        } else {
          const columnOrderCopy = [...columnOrder];
          const columnsCopy = [...columns];
          while (columnsCopy.length && columnOrderCopy.length) {
            const targetColumnId = columnOrderCopy.shift();
            const foundIndex = columnsCopy.findIndex(
              (d) => d.id === targetColumnId
            );
            if (foundIndex > -1) {
              orderedColumns.push(columnsCopy.splice(foundIndex, 1)[0]);
            }
          }
          orderedColumns = [...orderedColumns, ...columnsCopy];
        }
        return orderColumns(orderedColumns, grouping, groupedColumnMode);
      },
      getMemoOptions(table.options, "debugTable", "_getOrderColumnsFn")
    );
  }
};
export {
  ColumnOrdering
};
