// @dynatrace/strato-components/esm/tables/DataTable/tanstack-table-core-override/src/features/GlobalFiltering.js
import { filterFns } from "../filterFns";
import { isFunction, makeStateUpdater } from "../utils";
const GlobalFiltering = {
  getInitialState: (state) => {
    return {
      globalFilter: void 0,
      ...state
    };
  },
  getDefaultOptions: (table) => {
    return {
      onGlobalFilterChange: makeStateUpdater("globalFilter", table),
      globalFilterFn: "auto",
      getColumnCanGlobalFilter: (column) => {
        const value = table.getCoreRowModel().flatRows[0]?._getAllCellsByColumnId()[column.id]?.getValue();
        return typeof value === "string" || typeof value === "number";
      }
    };
  },
  createColumn: (column, table) => {
    column.getCanGlobalFilter = () => {
      return (column.columnDef.enableGlobalFilter ?? true) && (table.options.enableGlobalFilter ?? true) && (table.options.enableFilters ?? true) && (table.options.getColumnCanGlobalFilter?.(column) ?? true) && !!column.accessorFn;
    };
  },
  createTable: (table) => {
    table.getGlobalAutoFilterFn = () => {
      return filterFns.includesString;
    };
    table.getGlobalFilterFn = () => {
      const { globalFilterFn } = table.options;
      return isFunction(globalFilterFn) ? globalFilterFn : globalFilterFn === "auto" ? table.getGlobalAutoFilterFn() : table.options.filterFns?.[globalFilterFn] ?? filterFns[globalFilterFn];
    };
    table.setGlobalFilter = (updater) => {
      table.options.onGlobalFilterChange?.(updater);
    };
    table.resetGlobalFilter = (defaultState) => {
      table.setGlobalFilter(
        defaultState ? void 0 : table.initialState.globalFilter
      );
    };
  }
};
export {
  GlobalFiltering
};
