// @dynatrace/strato-components/esm/tables/DataTable/tanstack-table-core-override/src/features/RowPinning.js
import { getRowProto } from "..";
import { getMemoOptions, makeStateUpdater, memo } from "../utils";
const getDefaultRowPinningState = () => ({
  top: [],
  bottom: []
});
const RowPinning = {
  getInitialState: (state) => {
    return {
      rowPinning: getDefaultRowPinningState(),
      ...state
    };
  },
  getDefaultOptions: (table) => {
    return {
      onRowPinningChange: makeStateUpdater("rowPinning", table)
    };
  },
  createTable: (table) => {
    table.setRowPinning = (updater) => table.options.onRowPinningChange?.(updater);
    table.resetRowPinning = (defaultState) => table.setRowPinning(
      defaultState ? getDefaultRowPinningState() : table.initialState?.rowPinning ?? getDefaultRowPinningState()
    );
    table.getIsSomeRowsPinned = (position) => {
      const pinningState = table.getState().rowPinning;
      if (!position) {
        return Boolean(pinningState.top?.length || pinningState.bottom?.length);
      }
      return Boolean(pinningState[position]?.length);
    };
    table._getPinnedRows = (visibleRows, pinnedRowIds, position) => {
      const rows = table.options.keepPinnedRows ?? true ? (
        //get all rows that are pinned even if they would not be otherwise visible
        //account for expanded parent rows, but not pagination or filtering
        (pinnedRowIds ?? []).map((rowId) => {
          const row = table.getRow(rowId, true);
          return row.getIsAllParentsExpanded() ? row : null;
        })
      ) : (
        //else get only visible rows that are pinned
        (pinnedRowIds ?? []).map(
          (rowId) => visibleRows.find((row) => row.id === rowId)
        )
      );
      return rows.filter(Boolean).map((d) => ({ ...d, position }));
    };
    table.getTopRows = memo(
      () => [table.getRowModel().rows, table.getState().rowPinning.top],
      (allRows, topPinnedRowIds) => table._getPinnedRows(allRows, topPinnedRowIds, "top"),
      getMemoOptions(table.options, "debugRows", "getTopRows")
    );
    table.getBottomRows = memo(
      () => [table.getRowModel().rows, table.getState().rowPinning.bottom],
      (allRows, bottomPinnedRowIds) => table._getPinnedRows(allRows, bottomPinnedRowIds, "bottom"),
      getMemoOptions(table.options, "debugRows", "getBottomRows")
    );
    table.getCenterRows = memo(
      () => [
        table.getRowModel().rows,
        table.getState().rowPinning.top,
        table.getState().rowPinning.bottom
      ],
      (allRows, top, bottom) => {
        const topAndBottom = /* @__PURE__ */ new Set([...top ?? [], ...bottom ?? []]);
        return allRows.filter((d) => !topAndBottom.has(d.id));
      },
      getMemoOptions(table.options, "debugRows", "getCenterRows")
    );
    Object.assign(getRowProto(table), {
      pin(position, includeLeafRows, includeParentRows) {
        const leafRowIds = includeLeafRows ? this.getLeafRows().map(({ id }) => id) : [];
        const parentRowIds = includeParentRows ? this.getParentRows().map(({ id }) => id) : [];
        const rowIds = /* @__PURE__ */ new Set([...parentRowIds, this.id, ...leafRowIds]);
        table.setRowPinning((old) => {
          if (position === "bottom") {
            return {
              top: (old?.top ?? []).filter((d) => !rowIds?.has(d)),
              bottom: [
                ...(old?.bottom ?? []).filter((d) => !rowIds?.has(d)),
                ...Array.from(rowIds)
              ]
            };
          }
          if (position === "top") {
            return {
              top: [
                ...(old?.top ?? []).filter((d) => !rowIds?.has(d)),
                ...Array.from(rowIds)
              ],
              bottom: (old?.bottom ?? []).filter((d) => !rowIds?.has(d))
            };
          }
          return {
            top: (old?.top ?? []).filter((d) => !rowIds?.has(d)),
            bottom: (old?.bottom ?? []).filter((d) => !rowIds?.has(d))
          };
        });
      },
      getCanPin() {
        const { enableRowPinning, enablePinning } = table.options;
        if (typeof enableRowPinning === "function") {
          return enableRowPinning(this);
        }
        return enableRowPinning ?? enablePinning ?? true;
      },
      getIsPinned() {
        const rowIds = [this.id];
        const { top, bottom } = table.getState().rowPinning;
        const isTop = rowIds.some((d) => top?.includes(d));
        const isBottom = rowIds.some((d) => bottom?.includes(d));
        return isTop ? "top" : isBottom ? "bottom" : false;
      },
      getPinnedIndex() {
        const position = this.getIsPinned();
        if (!position) return -1;
        const visiblePinnedRowIds = (position === "top" ? table.getTopRows() : table.getBottomRows())?.map(({ id }) => id);
        return visiblePinnedRowIds?.indexOf(this.id) ?? -1;
      }
    });
  }
};
export {
  RowPinning
};
