// @dynatrace/strato-components/esm/tables/DataTable/tanstack-table-core-override/src/features/RowSelection.js
import { getRowProto } from "..";
import { getMemoOptions, makeStateUpdater, memo } from "../utils";
const RowSelection = {
  getInitialState: (state) => {
    return {
      rowSelection: {},
      ...state
    };
  },
  getDefaultOptions: (table) => {
    return {
      onRowSelectionChange: makeStateUpdater("rowSelection", table),
      enableRowSelection: true,
      enableMultiRowSelection: true,
      enableSubRowSelection: true
      // enableGroupingRowSelection: false,
      // isAdditiveSelectEvent: (e: unknown) => !!e.metaKey,
      // isInclusiveSelectEvent: (e: unknown) => !!e.shiftKey,
    };
  },
  createTable: (table) => {
    table.setRowSelection = (updater) => table.options.onRowSelectionChange?.(updater);
    table.resetRowSelection = (defaultState) => table.setRowSelection(
      defaultState ? {} : table.initialState.rowSelection ?? {}
    );
    table.toggleAllRowsSelected = (value) => {
      table.setRowSelection((old) => {
        value = typeof value !== "undefined" ? value : !table.getIsAllRowsSelected();
        const rowSelection = { ...old };
        const preGroupedFlatRows = table.getPreGroupedRowModel().flatRows;
        if (value) {
          preGroupedFlatRows.forEach((row) => {
            if (!row.getCanSelect()) {
              return;
            }
            rowSelection[row.id] = true;
          });
        } else {
          preGroupedFlatRows.forEach((row) => {
            delete rowSelection[row.id];
          });
        }
        return rowSelection;
      });
    };
    table.toggleAllPageRowsSelected = (value) => table.setRowSelection((old) => {
      const resolvedValue = typeof value !== "undefined" ? value : !table.getIsAllPageRowsSelected();
      const rowSelection = { ...old };
      table.getRowModel().rows.forEach((row) => {
        mutateRowIsSelected(rowSelection, row.id, resolvedValue, true, table);
      });
      return rowSelection;
    });
    table.getPreSelectedRowModel = () => table.getCoreRowModel();
    table.getSelectedRowModel = memo(
      () => [table.getState().rowSelection, table.getCoreRowModel()],
      (rowSelection, rowModel) => {
        if (!Object.keys(rowSelection).length) {
          return {
            rows: [],
            flatRows: [],
            rowsById: {}
          };
        }
        return selectRowsFn(table, rowModel);
      },
      getMemoOptions(table.options, "debugTable", "getSelectedRowModel")
    );
    table.getFilteredSelectedRowModel = memo(
      () => [table.getState().rowSelection, table.getFilteredRowModel()],
      (rowSelection, rowModel) => {
        if (!Object.keys(rowSelection).length) {
          return {
            rows: [],
            flatRows: [],
            rowsById: {}
          };
        }
        return selectRowsFn(table, rowModel);
      },
      getMemoOptions(
        table.options,
        "debugTable",
        "getFilteredSelectedRowModel"
      )
    );
    table.getGroupedSelectedRowModel = memo(
      () => [table.getState().rowSelection, table.getSortedRowModel()],
      (rowSelection, rowModel) => {
        if (!Object.keys(rowSelection).length) {
          return {
            rows: [],
            flatRows: [],
            rowsById: {}
          };
        }
        return selectRowsFn(table, rowModel);
      },
      getMemoOptions(table.options, "debugTable", "getGroupedSelectedRowModel")
    );
    table.getIsAllRowsSelected = () => {
      const preGroupedFlatRows = table.getFilteredRowModel().flatRows;
      const { rowSelection } = table.getState();
      let isAllRowsSelected = Boolean(
        preGroupedFlatRows.length && Object.keys(rowSelection).length
      );
      if (isAllRowsSelected) {
        if (preGroupedFlatRows.some(
          (row) => row.getCanSelect() && !rowSelection[row.id]
        )) {
          isAllRowsSelected = false;
        }
      }
      return isAllRowsSelected;
    };
    table.getIsAllPageRowsSelected = () => {
      const paginationFlatRows = table.getPaginationRowModel().flatRows.filter((row) => row.getCanSelect());
      const { rowSelection } = table.getState();
      let isAllPageRowsSelected = !!paginationFlatRows.length;
      if (isAllPageRowsSelected && paginationFlatRows.some((row) => !rowSelection[row.id])) {
        isAllPageRowsSelected = false;
      }
      return isAllPageRowsSelected;
    };
    table.getIsSomeRowsSelected = () => {
      const totalSelected = Object.keys(
        table.getState().rowSelection ?? {}
      ).length;
      return totalSelected > 0 && totalSelected < table.getFilteredRowModel().flatRows.length;
    };
    table.getIsSomePageRowsSelected = () => {
      const paginationFlatRows = table.getPaginationRowModel().flatRows;
      return table.getIsAllPageRowsSelected() ? false : paginationFlatRows.filter((row) => row.getCanSelect()).some((d) => d.getIsSelected() || d.getIsSomeSelected());
    };
    table.getToggleAllRowsSelectedHandler = () => {
      return (e) => {
        table.toggleAllRowsSelected(
          e.target.checked
        );
      };
    };
    table.getToggleAllPageRowsSelectedHandler = () => {
      return (e) => {
        table.toggleAllPageRowsSelected(
          e.target.checked
        );
      };
    };
    Object.assign(getRowProto(table), {
      toggleSelected(value, opts) {
        const isSelected = this.getIsSelected();
        table.setRowSelection((old) => {
          value = typeof value !== "undefined" ? value : !isSelected;
          if (this.getCanSelect() && isSelected === value) {
            return old;
          }
          const selectedRowIds = { ...old };
          mutateRowIsSelected(
            selectedRowIds,
            this.id,
            value,
            opts?.selectChildren ?? true,
            table
          );
          return selectedRowIds;
        });
      },
      getIsSelected() {
        const { rowSelection } = table.getState();
        return isRowSelected(this, rowSelection);
      },
      getIsSomeSelected() {
        const { rowSelection } = table.getState();
        return isSubRowSelected(this, rowSelection, table) === "some";
      },
      getIsAllSubRowsSelected() {
        const { rowSelection } = table.getState();
        return isSubRowSelected(this, rowSelection, table) === "all";
      },
      getCanSelect() {
        if (typeof table.options.enableRowSelection === "function") {
          return table.options.enableRowSelection(this);
        }
        return table.options.enableRowSelection ?? true;
      },
      getCanSelectSubRows() {
        if (typeof table.options.enableSubRowSelection === "function") {
          return table.options.enableSubRowSelection(this);
        }
        return table.options.enableSubRowSelection ?? true;
      },
      getCanMultiSelect() {
        if (typeof table.options.enableMultiRowSelection === "function") {
          return table.options.enableMultiRowSelection(this);
        }
        return table.options.enableMultiRowSelection ?? true;
      },
      getToggleSelectedHandler() {
        const canSelect = this.getCanSelect();
        return (e) => {
          if (!canSelect) return;
          this.toggleSelected(
            e.target?.checked
          );
        };
      }
    });
  }
};
const mutateRowIsSelected = (selectedRowIds, id, value, includeChildren, table) => {
  const row = table.getRow(id, true);
  if (value) {
    if (!row.getCanMultiSelect()) {
      Object.keys(selectedRowIds).forEach((key) => delete selectedRowIds[key]);
    }
    if (row.getCanSelect()) {
      selectedRowIds[id] = true;
    }
  } else {
    delete selectedRowIds[id];
  }
  if (includeChildren && row.subRows?.length && row.getCanSelectSubRows()) {
    row.subRows.forEach(
      (row2) => mutateRowIsSelected(
        selectedRowIds,
        row2.id,
        value,
        includeChildren,
        table
      )
    );
  }
};
function selectRowsFn(table, rowModel) {
  const rowSelection = table.getState().rowSelection;
  const newSelectedFlatRows = [];
  const newSelectedRowsById = {};
  const recurseRows = (rows, depth = 0) => {
    return rows.map((row) => {
      const isSelected = isRowSelected(row, rowSelection);
      if (isSelected) {
        newSelectedFlatRows.push(row);
        newSelectedRowsById[row.id] = row;
      }
      if (row.subRows?.length) {
        row = row.clone();
        row.subRows = recurseRows(row.subRows, depth + 1);
      }
      if (isSelected) {
        return row;
      }
    }).filter(Boolean);
  };
  return {
    rows: recurseRows(rowModel.rows),
    flatRows: newSelectedFlatRows,
    rowsById: newSelectedRowsById
  };
}
function isRowSelected(row, selection) {
  return selection[row.id] ?? false;
}
function isSubRowSelected(row, selection, table) {
  if (!row.subRows?.length) return false;
  let allChildrenSelected = true;
  let someSelected = false;
  row.subRows.forEach((subRow) => {
    if (someSelected && !allChildrenSelected) {
      return;
    }
    if (subRow.getCanSelect()) {
      if (isRowSelected(subRow, selection)) {
        someSelected = true;
      } else {
        allChildrenSelected = false;
      }
    }
    if (subRow.subRows && subRow.subRows.length) {
      const subRowChildrenSelected = isSubRowSelected(subRow, selection, table);
      if (subRowChildrenSelected === "all") {
        someSelected = true;
      } else if (subRowChildrenSelected === "some") {
        someSelected = true;
        allChildrenSelected = false;
      } else {
        allChildrenSelected = false;
      }
    }
  });
  return allChildrenSelected ? "all" : someSelected ? "some" : false;
}
export {
  RowSelection,
  isRowSelected,
  isSubRowSelected,
  selectRowsFn
};
