// @dynatrace/strato-components/editors/code-editor/languages/languages.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var languages_exports = {};
__export(languages_exports, {
  CodeEditorLanguageKeys: () => CodeEditorLanguageKeys,
  codeEditorLanguages: () => codeEditorLanguages
});
module.exports = __toCommonJS(languages_exports);
var import_lang_javascript = require("@codemirror/lang-javascript");
var import_lang_json = require("@codemirror/lang-json");
var import_lang_markdown = require("@codemirror/lang-markdown");
var import_language = require("@codemirror/language");
var import_yaml = require("@codemirror/legacy-modes/mode/yaml");
var import_lint = require("@codemirror/lint");
var import_javascript = require("../auto-completion/javascript.js");
const codeEditorLanguages = {
  other: {
    extendWith: []
  },
  json: {
    language: (0, import_lang_json.json)().language,
    extendWith: [(0, import_lint.linter)((0, import_lang_json.jsonParseLinter)())]
  },
  js: {
    language: (0, import_lang_javascript.javascript)().language,
    extendWith: [],
    autocompleteOptions: {
      activateOnTyping: true,
      providers: [import_javascript.javascriptAutocompleteProvider]
    }
  },
  jsx: {
    language: (0, import_lang_javascript.javascript)({ jsx: true }).language,
    extendWith: [],
    autocompleteOptions: {
      activateOnTyping: true,
      providers: [import_javascript.javascriptAutocompleteProvider]
    }
  },
  ts: {
    language: (0, import_lang_javascript.javascript)({ typescript: true }).language,
    extendWith: [],
    autocompleteOptions: {
      activateOnTyping: true,
      providers: [import_javascript.javascriptAutocompleteProvider]
    }
  },
  tsx: {
    language: (0, import_lang_javascript.javascript)({ typescript: true, jsx: true }).language,
    extendWith: [],
    autocompleteOptions: {
      activateOnTyping: true,
      providers: [import_javascript.javascriptAutocompleteProvider]
    }
  },
  md: {
    language: (0, import_lang_markdown.markdown)().language,
    extendWith: []
  },
  yaml: {
    language: import_language.StreamLanguage.define(import_yaml.yaml),
    extendWith: []
  }
};
const CodeEditorLanguageKeys = Object.keys(
  codeEditorLanguages
);
