// @dynatrace/strato-components/esm/charts/core/components/chart-grid/ChartGridRoot.js
import { jsx } from "react/jsx-runtime";
import { forwardRef, useRef } from "react";
import { ChartGridArea } from "./ChartGridArea.js";
import { ChartGridPlot } from "./ChartGridPlot.js";
import { GridPlotSizeProvider } from "./providers/GridPlotSizeProvider.js";
import { useMergeRefs } from "../../../../core/hooks/useMergeRefs.js";
import { Grid } from "../../../../layouts/grid/Grid.js";
const ChartGridRoot = /* @__PURE__ */ forwardRef(
  ({
    width,
    height,
    gridTemplateAreas,
    gridTemplateColumns,
    gridTemplateRows,
    children,
    ...props
  }, forwardedRef) => {
    const chartRef = useRef(null);
    const mergedRef = useMergeRefs([forwardedRef, chartRef]);
    return /* @__PURE__ */ jsx(
      Grid,
      {
        ref: mergedRef,
        width,
        height,
        gridTemplateAreas,
        gridTemplateColumns,
        gridTemplateRows,
        minHeight: "min-content",
        gap: 0,
        ...props,
        children: /* @__PURE__ */ jsx(GridPlotSizeProvider, { children })
      }
    );
  }
);
ChartGridRoot.displayName = "ChartGridRoot";
const ChartGrid = Object.assign(ChartGridRoot, {
  PlotArea: ChartGridPlot,
  GridArea: ChartGridArea
});
export {
  ChartGrid,
  ChartGridRoot
};
