// @dynatrace/strato-components/tables/DataTable/features/MeterBarChart/MeterBarChart.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MeterBarChart_exports = {};
__export(MeterBarChart_exports, {
  MeterBarChartRenderer: () => MeterBarChartRenderer,
  getTotalMaxFromMeterbarChartData: () => getTotalMaxFromMeterbarChartData,
  isMultiMeterBarData: () => isMultiMeterBarData,
  useMeterBarChartMaxValuesCalc: () => useMeterBarChartMaxValuesCalc
});
module.exports = __toCommonJS(MeterBarChart_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_spacings = __toESM(require("@dynatrace/strato-design-tokens/spacings"));
var import_MultiMeterBarChart = require("../../../../charts/meter-bar/multi-meter-bar/MultiMeterBarChart.js");
var import_constants = require("../../../../charts/meter-bar/shared/constants.js");
var import_MeterBarChart = require("../../../../charts/meter-bar/single-meter-bar/MeterBarChart.js");
var import_logging = require("../../../../core/utils/logging.js");
var import_DataTableContext = require("../../contexts/DataTableContext.js");
var import_datatable_typeguards = require("../../hooks/useTable/datatable-typeguards.js");
const DEFAULT_MAX_VALUE = 100;
function isDataTableMeterBarColumnConfig(config) {
  return config !== null && config !== void 0 && typeof config === "object" && ("max" in config || "color" in config || "threshold" in config || "showTooltip" in config || "max" in config);
}
function isColumnDefWithAccessor(column) {
  return column !== null && column !== void 0 && typeof column === "object" && "accessor" in column;
}
function isMultiMeterBarData(value) {
  return Array.isArray(value) && value.every(
    (item) => typeof item === "object" && typeof item.name === "string" && typeof item.value === "number" && (item.color === void 0 || typeof item.color === "string")
  );
}
function MeterBarChartRenderer({
  value,
  column
}) {
  const table = (0, import_DataTableContext.useDataTable)();
  const chartRef = (0, import_react.useRef)(null);
  let config = {};
  if (column?.columnDef.meta && isDataTableMeterBarColumnConfig(column.columnDef.meta.config)) {
    config = column.columnDef.meta.config;
  }
  const columnId = column.columnDef.id ?? "";
  const maxValuesFromState = table.getState().meterbarChartMaxValues;
  const maxValue = !(0, import_lodash_es.isEmpty)(maxValuesFromState) ? maxValuesFromState[columnId] : DEFAULT_MAX_VALUE;
  const listenerRef = (0, import_react.useRef)(null);
  (0, import_react.useEffect)(() => {
    return () => {
      listenerRef.current?.();
    };
  }, []);
  const registerUnpinOnPointerUp = config?.showTooltip ? () => {
    if (listenerRef.current) {
      return;
    }
    const handler = (event) => {
      if (event.target.closest('[role="tooltip"]') !== null) {
        return;
      }
      chartRef.current?.unpinTooltip();
      listenerRef.current?.();
      listenerRef.current = null;
    };
    listenerRef.current = () => document.removeEventListener("pointerup", handler);
    document.addEventListener("pointerup", handler);
  } : void 0;
  if (typeof value === "number") {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "span",
      {
        style: { display: "contents" },
        onClick: registerUnpinOnPointerUp,
        role: "none",
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
          import_MeterBarChart.MeterBarChart,
          {
            value,
            min: config?.min,
            max: maxValue,
            color: config?.color,
            size: import_constants.METER_BAR_SIZE.Size8,
            width: "100%",
            height: "auto",
            style: {
              pointerEvents: "auto",
              padding: `${import_spacings.default.Size6} 0`
            },
            formatter: config?.formatter,
            ref: chartRef,
            children: [
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_MeterBarChart.MeterBarChart.Tooltip, { hidden: !config?.showTooltip }),
              config?.thresholds?.map((threshold) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_MeterBarChart.MeterBarChart.ThresholdIndicator,
                {
                  name: threshold.name,
                  color: threshold.color,
                  value: threshold.value,
                  showIndicator: threshold.showIndicator
                },
                `threshold-${threshold.color}-${threshold.value}`
              ))
            ]
          }
        )
      }
    );
  }
  const isMultiMeterData = isMultiMeterBarData(value);
  if (!isMultiMeterData) {
    (0, import_logging.devLog)("Invalid meter chart data.");
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "span",
    {
      style: { display: "contents" },
      onClick: registerUnpinOnPointerUp,
      role: "none",
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
        import_MultiMeterBarChart.MultiMeterBarChart,
        {
          size: "size8",
          min: config?.min,
          max: maxValue,
          colorPalette: config?.colorPalette || "categorical",
          width: "100%",
          height: "auto",
          style: {
            pointerEvents: "auto",
            padding: `${import_spacings.default.Size6} 0`
          },
          formatter: config?.formatter,
          ref: chartRef,
          // only if this is multi meter data render segments
          children: [
            isMultiMeterData && value.map((data) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_MultiMeterBarChart.MultiMeterBarChart.Segment,
              {
                name: data.name,
                value: data.value,
                color: data?.color
              },
              `${data.name}-${data.value}`
            )),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_MultiMeterBarChart.MultiMeterBarChart.Tooltip,
              {
                variant: "shared",
                hidden: !config?.showTooltip
              }
            )
          ]
        }
      )
    }
  );
}
function getTotalMaxFromMeterbarChartData(data, columns) {
  const meterbarColumns = columns.filter(
    (columnDef) => (0, import_datatable_typeguards.isDisplayColumnDef)(columnDef) && columnDef.columnType === "meterbar"
  );
  const maxValuesStorage = {};
  for (const row of data) {
    for (const columnDef of meterbarColumns) {
      if (isColumnDefWithAccessor(columnDef)) {
        const { accessor, config, id: columnId } = columnDef;
        const configuredMax = isDataTableMeterBarColumnConfig(config) ? config.max : void 0;
        const meterbarValue = row;
        const meterbarDataEntry = typeof accessor === "string" ? meterbarValue[accessor] : accessor(row);
        let totalValue = DEFAULT_MAX_VALUE;
        if ((0, import_lodash_es.isNumber)(configuredMax)) {
          maxValuesStorage[columnId] = configuredMax;
          continue;
        }
        if (meterbarDataEntry) {
          if (Array.isArray(meterbarDataEntry)) {
            totalValue = meterbarDataEntry.reduce(
              (acc, current) => acc + current.value,
              0
            );
          } else {
            const entry = (0, import_lodash_es.isNumber)(meterbarDataEntry) ? meterbarDataEntry : DEFAULT_MAX_VALUE;
            totalValue = configuredMax === "data-max" ? entry : Math.max(entry, DEFAULT_MAX_VALUE);
          }
        }
        maxValuesStorage[columnId] = Math.max(
          maxValuesStorage[columnId] ?? totalValue,
          totalValue
        );
      }
    }
  }
  return maxValuesStorage;
}
function useMeterBarChartMaxValuesCalc(props, options) {
  const meterbarChartMaxValues = (0, import_react.useMemo)(
    () => getTotalMaxFromMeterbarChartData(props.data, props.columns),
    [props.columns, props.data]
  );
  options.state = { ...options.state, meterbarChartMaxValues };
}
