// @dynatrace/strato-components/navigation/menu/Content.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Content_exports = {};
__export(Content_exports, {
  Content: () => Content
});
module.exports = __toCommonJS(Content_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_dropdown_menu = require("@radix-ui/react-dropdown-menu");
var import_react = require("react");
var import_useMenuContext = require("./hooks/useMenuContext.js");
var styles = __toESM(require("./Menu.sty.js"));
var import_useOverlayStacking = require("../../core/hooks/useOverlayStacking.js");
const SIDE_OFFSET = 4;
const Content = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    side = "bottom",
    alignment = "start",
    onCloseAutoFocus,
    avoidCollisions = true,
    _sideOffset = SIDE_OFFSET,
    _onTabKeyPressed,
    children,
    collisionBoundary,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    ...remainingProps
  } = props;
  const tabKeyHandler = (0, import_react.useCallback)(
    (e) => {
      if (_onTabKeyPressed && e.key === "Tab") {
        _onTabKeyPressed();
      }
    },
    [_onTabKeyPressed]
  );
  const { style } = (0, import_useOverlayStacking.useOverlayStacking)(void 0, "overlay");
  const { setIsOpen } = (0, import_useMenuContext.useMenuContext)();
  const handleOnKeyDown = (0, import_react.useCallback)(
    (e) => {
      if (e.key === "Escape") {
        setIsOpen(false);
        e.stopPropagation();
      }
    },
    [setIsOpen]
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_dropdown_menu.Portal, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_react_dropdown_menu.Content,
    {
      ...remainingProps,
      "data-testid": dataTestId,
      ref: forwardedRef,
      side,
      align: alignment,
      onKeyUp: tabKeyHandler,
      onCloseAutoFocus,
      sideOffset: _sideOffset,
      avoidCollisions,
      className: styles.content,
      style: { zIndex: style.zIndex },
      onKeyDown: handleOnKeyDown,
      collisionBoundary,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      children
    }
  ) });
});
Content.displayName = "Menu.Content";
