// @dynatrace/strato-components/navigation/menu/Intent.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Intent_exports = {};
__export(Intent_exports, {
  Intent: () => Intent,
  getHandleSelect: () => getHandleSelect,
  getIntentIcon: () => getIntentIcon
});
module.exports = __toCommonJS(Intent_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_navigation = require("@dynatrace-sdk/navigation");
var import_item_slots = require("./item-slots.js");
var import_Item = require("./Item.js");
var import_Prefix = require("./Prefix.js");
var import_intent = require("../../core/hooks/intent.js");
var import_AppIcon = require("../../layouts/app-header/app-icon/AppIcon.js");
function isWithResponse(props) {
  if ("onResponse" in props) {
    return typeof props.onResponse === "function";
  }
  return false;
}
function getIntentIcon(icon, intentIcon) {
  if (!icon && !intentIcon) {
    return null;
  }
  if (icon) {
    return icon;
  }
  if (intentIcon) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Prefix.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_AppIcon.AppIcon, { src: intentIcon, alt: "" }) });
  }
}
const getHandleSelect = (intent, onSelect, onResponse) => async (e) => {
  onSelect?.(e);
  try {
    const response = await intent.send();
    onResponse(response);
  } catch {
    onResponse({
      status: "error",
      error: new Error("Error sending intent with response")
    });
  }
};
const IntentWithoutResponse = /* @__PURE__ */ (0, import_react.forwardRef)((props, ref) => {
  const {
    onResponse,
    onSelect,
    payload,
    options,
    children,
    disabled,
    ...rest
  } = props;
  const intent = (0, import_intent.useIntent)({
    payload,
    options
  });
  const slots = (0, import_item_slots.getIntentSlots)(children);
  const intentIcon = getIntentIcon(slots.prefix, intent.iconURL);
  const link = (0, import_react.useMemo)(() => {
    return (0, import_navigation.getIntentLink)(
      payload,
      options.recommendedAppId ?? "",
      options.recommendedIntentId ?? ""
    );
  }, [options.recommendedAppId, options.recommendedIntentId, payload]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_Item.Item,
    {
      ...rest,
      as: "a",
      href: disabled ? void 0 : link,
      ref,
      disabled,
      onSelect: (e) => {
        e.preventDefault();
        intent.send();
        onSelect?.(e);
      },
      children: [
        intentIcon,
        slots.freeform
      ]
    }
  );
});
const IntentWithResponse = /* @__PURE__ */ (0, import_react.forwardRef)((props, ref) => {
  const {
    onSelect,
    onResponse,
    payload,
    options,
    children,
    disabled,
    ...rest
  } = props;
  const intent = (0, import_intent.useIntentWithResponse)({
    payload,
    options
  });
  const slots = (0, import_item_slots.getIntentSlots)(children);
  const intentIcon = getIntentIcon(slots.prefix, intent.iconURL);
  const handleSelect = getHandleSelect(intent, onSelect, onResponse);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Item.Item, { ...rest, ref, disabled, onSelect: handleSelect, children: [
    intentIcon,
    slots.freeform
  ] });
});
const Intent = /* @__PURE__ */ (0, import_react.forwardRef)(
  (props, ref) => {
    if (isWithResponse(props)) {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(IntentWithResponse, { ...props, ref });
    }
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(IntentWithoutResponse, { ...props, ref });
  }
);
Intent.displayName = "Menu.Intent";
