// @dynatrace/strato-components/navigation/menu/ItemTooltip.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ItemTooltip_exports = {};
__export(ItemTooltip_exports, {
  ItemTooltip: () => ItemTooltip,
  MenuItemTooltipWrapper: () => MenuItemTooltipWrapper
});
module.exports = __toCommonJS(ItemTooltip_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_popper = require("react-popper");
var menuStyles = __toESM(require("./Menu.sty.js"));
var import_SubContent = require("./SubContent.js");
const MenuItemTooltipWrapper = ({
  triggerElement,
  children,
  onMouseEnter,
  onMouseLeave
}) => {
  const [menuTooltipRefRef, setMenuTooltipRefRef] = (0, import_react.useState)(null);
  const { styles, attributes } = (0, import_react_popper.usePopper)(triggerElement, menuTooltipRefRef, {
    placement: "right-start",
    modifiers: [
      {
        name: "offset",
        options: {
          offset: [import_SubContent.ALIGN_OFFSET, 0]
        }
      },
      {
        name: "flip",
        options: {
          fallbackPlacements: ["left-start"]
        }
      }
    ]
  });
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      ref: setMenuTooltipRefRef,
      style: styles.popper,
      className: menuStyles.tooltipWrapper,
      onMouseEnter,
      onMouseLeave,
      ...attributes.popper,
      children
    }
  );
};
MenuItemTooltipWrapper.displayName = "MenuItemTooltipWrapper";
const ItemTooltip = (0, import_react.forwardRef)(
  ({ children, "data-testid": dataTestId }, forwardedRef) => children ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      className: menuStyles.tooltip,
      ref: forwardedRef,
      "data-testid": dataTestId,
      children
    }
  ) : null
);
ItemTooltip.displayName = "Menu.ItemTooltip";
