// @dynatrace/strato-components/navigation/menu/item-slots.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var item_slots_exports = {};
__export(item_slots_exports, {
  getIntentSlots: () => getIntentSlots,
  getItemSlots: () => getItemSlots,
  getSubTriggerSlots: () => getSubTriggerSlots
});
module.exports = __toCommonJS(item_slots_exports);
var import_react = require("react");
var import_react_is = require("react-is");
var import_ItemTooltip = require("./ItemTooltip.js");
var import_Prefix = require("./Prefix.js");
var import_Suffix = require("./Suffix.js");
function getItemSlots(children) {
  const slots = {
    tooltip: null,
    prefix: null,
    suffix: null,
    freeform: []
  };
  import_react.Children.forEach(children, (child) => {
    if ((0, import_react_is.isElement)(child)) {
      switch (child.type) {
        case import_ItemTooltip.ItemTooltip:
          slots.tooltip = child;
          break;
        case import_Prefix.Prefix:
          slots.prefix = child;
          break;
        case import_Suffix.Suffix:
          slots.suffix = child;
          break;
        default:
          slots.freeform.push(child);
      }
    } else {
      slots.freeform.push(child);
    }
  });
  return slots;
}
function getSubTriggerSlots(children) {
  const slots = {
    prefix: null,
    freeform: []
  };
  import_react.Children.forEach(children, (child) => {
    if ((0, import_react_is.isElement)(child)) {
      switch (child.type) {
        case import_ItemTooltip.ItemTooltip:
          break;
        case import_Prefix.Prefix:
          slots.prefix = child;
          break;
        case import_Suffix.Suffix:
          break;
        default:
          slots.freeform.push(child);
      }
    } else {
      slots.freeform.push(child);
    }
  });
  return slots;
}
const getIntentSlots = getSubTriggerSlots;
