// @dynatrace/strato-components/navigation/menu/Menu.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Menu_exports = {};
__export(Menu_exports, {
  Menu: () => Menu
});
module.exports = __toCommonJS(Menu_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_dropdown_menu = require("@radix-ui/react-dropdown-menu");
var import_react = require("react");
var import_react_is = require("react-is");
var import_constants = require("./constants.js");
var import_Content = require("./Content.js");
var import_MenuContext = require("./contexts/MenuContext.js");
var import_Group = require("./Group.js");
var import_useMenuAutoClose = require("./hooks/useMenuAutoClose.js");
var import_Intent = require("./Intent.js");
var import_Item = require("./Item.js");
var import_ItemTooltip = require("./ItemTooltip.js");
var import_Label = require("./Label.js");
var import_Link = require("./Link.js");
var import_Prefix = require("./Prefix.js");
var import_Sub = require("./Sub.js");
var import_SubContent = require("./SubContent.js");
var import_SubTrigger = require("./SubTrigger.js");
var import_Suffix = require("./Suffix.js");
var import_Trigger = require("./Trigger.js");
var import_useControllableState = require("../../core/hooks/useControllableState.js");
var import_merge_props = require("../../core/utils/merge-props.js");
function getMenuSlots(children) {
  const slots = {
    trigger: null,
    content: [],
    freeform: []
  };
  import_react.Children.forEach(children, (child) => {
    if ((0, import_react_is.isElement)(child)) {
      switch (child.type) {
        case import_Trigger.Trigger:
          slots.trigger = child;
          break;
        case import_Content.Content:
          slots.content.push(child);
          break;
        default:
          slots.freeform.push(child);
      }
    } else {
      slots.freeform.push(child);
    }
  });
  return slots;
}
function registerTabHandlers(children, handleTabKeyPressed, dataDtComponent) {
  return import_react.Children.map(children, (child) => {
    return (0, import_react.isValidElement)(child) && child.type === import_Content.Content ? (0, import_react.cloneElement)(
      child,
      (0, import_merge_props.mergeProps)(child.props, {
        _onTabKeyPressed: handleTabKeyPressed,
        "data-dt-component": dataDtComponent
      }),
      child.props.children
    ) : child;
  });
}
const MenuRoot = (props) => {
  const {
    open: value,
    onOpenChange: onChange,
    modal = true,
    defaultOpen = false,
    children,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    // @ts-expect-error - We want to allow overriding data-dt-component as we use the component internally in other Strato components, but it should not appear in the public API.
    "data-dt-component": dataDtComponent = import_constants.COMPONENT_NAME
  } = props;
  const [open, updateOpen] = (0, import_useControllableState.useControllableState)({
    defaultValue: defaultOpen,
    value,
    onChange
  });
  const slots = (0, import_react.useMemo)(
    () => getMenuSlots(
      registerTabHandlers(children, () => updateOpen(false), dataDtComponent)
    ),
    [children, updateOpen, dataDtComponent]
  );
  (0, import_useMenuAutoClose.useMenuAutoClose)(open, updateOpen);
  const menuContextValue = (0, import_react.useMemo)(
    () => ({
      isOpen: open,
      isDefaultOpen: defaultOpen,
      setIsOpen: updateOpen
    }),
    [defaultOpen, open, updateOpen]
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_MenuContext.MenuContext.Provider, { value: menuContextValue, children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_react_dropdown_menu.Root,
    {
      open,
      onOpenChange: (isOpen) => updateOpen(isOpen),
      modal,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      children: [
        slots.trigger !== null && slots.trigger,
        open && slots.content,
        slots.freeform
      ]
    }
  ) });
};
MenuRoot.displayName = "Menu";
const Menu = Object.assign(MenuRoot, {
  Trigger: import_Trigger.Trigger,
  Content: import_Content.Content,
  Item: import_Item.Item,
  ItemTooltip: import_ItemTooltip.ItemTooltip,
  Label: import_Label.Label,
  Link: import_Link.Link,
  Group: import_Group.Group,
  Prefix: import_Prefix.Prefix,
  Suffix: import_Suffix.Suffix,
  Sub: import_Sub.Sub,
  SubTrigger: import_SubTrigger.SubTrigger,
  SubContent: import_SubContent.SubContent,
  Intent: import_Intent.Intent
});
