// @dynatrace/strato-components/navigation/menu/Menu.sty.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Menu_css_exports = {};
__export(Menu_css_exports, {
  content: () => content,
  group: () => group,
  item: () => item,
  itemChildren: () => itemChildren,
  label: () => label,
  subChildren: () => subChildren,
  tooltip: () => tooltip,
  tooltipWrapper: () => tooltipWrapper
});
module.exports = __toCommonJS(Menu_css_exports);
var import_Menu_css_ts_vanilla = require("./Menu.css");
var import_createRuntimeFn = require("@vanilla-extract/recipes/createRuntimeFn");
var content = "_1m01arl3-3-6-0";
var group = "_1m01arl4-3-6-0";
var item = (0, import_createRuntimeFn.createRuntimeFn)({ defaultClassName: "_1m01arl5-3-6-0", variantClassNames: { disabled: { true: "_1m01arl6-3-6-0", false: "_1m01arl7-3-6-0" } }, defaultVariants: { disabled: false }, compoundVariants: [] });
var itemChildren = "_1m01arl8-3-6-0";
var label = "_1m01arl2-3-6-0";
var subChildren = "_1m01arl9-3-6-0";
var tooltip = "_1m01arl0-3-6-0";
var tooltipWrapper = "_1m01arl1-3-6-0";
