// @dynatrace/strato-components/navigation/menu/Sub.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Sub_exports = {};
__export(Sub_exports, {
  Sub: () => Sub
});
module.exports = __toCommonJS(Sub_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_dropdown_menu = require("@radix-ui/react-dropdown-menu");
var import_react = require("react");
var import_react_is = require("react-is");
var import_MenuSubContext = require("./contexts/MenuSubContext.js");
var import_useMenuAutoClose = require("./hooks/useMenuAutoClose.js");
var import_SubContent = require("./SubContent.js");
var import_SubTrigger = require("./SubTrigger.js");
var import_useControllableState = require("../../core/hooks/useControllableState.js");
var import_merge_props = require("../../core/utils/merge-props.js");
function registerTabHandlers(children, handleTabKeyPressed) {
  return import_react.Children.map(children, (child) => {
    return (0, import_react.isValidElement)(child) && child.type === import_SubContent.SubContent ? (0, import_react.cloneElement)(
      child,
      (0, import_merge_props.mergeProps)(child.props, {
        _onTabKeyPressed: handleTabKeyPressed
      }),
      child.props.children
    ) : child;
  });
}
function getSubSlots(children) {
  const slots = {
    trigger: null,
    content: [],
    freeform: []
  };
  import_react.Children.forEach(children, (child) => {
    if ((0, import_react_is.isElement)(child)) {
      switch (child.type) {
        case import_SubTrigger.SubTrigger:
          slots.trigger = child;
          break;
        case import_SubContent.SubContent:
          slots.content.push(child);
          break;
        default:
          slots.freeform.push(child);
      }
    } else {
      slots.freeform.push(child);
    }
  });
  return slots;
}
const Sub = (props) => {
  const {
    defaultOpen: defaultValue,
    open: value,
    onOpenChange: onChange,
    children,
    ...remainingProps
  } = props;
  const [open, updateOpen] = (0, import_useControllableState.useControllableState)({
    defaultValue,
    value,
    onChange
  });
  const slots = (0, import_react.useMemo)(
    () => getSubSlots(registerTabHandlers(children, () => updateOpen(false))),
    [children, updateOpen]
  );
  (0, import_useMenuAutoClose.useMenuAutoClose)(open, updateOpen);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_MenuSubContext.MenuSubContext.Provider,
    {
      value: {
        isOpen: open ?? false,
        isDefaultOpen: defaultValue ?? false,
        setIsOpen: updateOpen
      },
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
        import_react_dropdown_menu.Sub,
        {
          open,
          onOpenChange: (isOpen) => updateOpen(isOpen),
          ...remainingProps,
          children: [
            slots.trigger !== null && slots.trigger,
            open && slots.content,
            slots.freeform
          ]
        }
      )
    }
  );
};
Sub.displayName = "Menu.Sub";
