// @dynatrace/strato-components/navigation/menu/SubContent.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SubContent_exports = {};
__export(SubContent_exports, {
  ALIGN_OFFSET: () => ALIGN_OFFSET,
  SubContent: () => SubContent
});
module.exports = __toCommonJS(SubContent_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_dropdown_menu = require("@radix-ui/react-dropdown-menu");
var import_react = require("react");
var styles = __toESM(require("./Menu.sty.js"));
var import_useOverlayStacking = require("../../core/hooks/useOverlayStacking.js");
const SUB_CONTENT_SIDE_OFFSET = 0;
const ALIGN_OFFSET = -4;
const SubContent = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const { _onTabKeyPressed, children } = props;
  const tabKeyHandler = (0, import_react.useCallback)(
    (e) => {
      if (_onTabKeyPressed && e.key === "Tab") {
        _onTabKeyPressed();
      }
    },
    [_onTabKeyPressed]
  );
  const { style } = (0, import_useOverlayStacking.useOverlayStacking)(void 0, "overlay");
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_dropdown_menu.Portal, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_react_dropdown_menu.SubContent,
    {
      ref: forwardedRef,
      sideOffset: SUB_CONTENT_SIDE_OFFSET,
      alignOffset: ALIGN_OFFSET,
      onKeyUp: tabKeyHandler,
      className: styles.content,
      style: { zIndex: style.zIndex },
      collisionPadding: 2,
      children
    }
  ) });
});
SubContent.displayName = "Menu.SubContent";
