// @dynatrace/strato-components/navigation/menu/SubTrigger.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SubTrigger_exports = {};
__export(SubTrigger_exports, {
  SubTrigger: () => SubTrigger
});
module.exports = __toCommonJS(SubTrigger_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_dropdown_menu = require("@radix-ui/react-dropdown-menu");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_useIsMenuItemHovered = require("./hooks/useIsMenuItemHovered.js");
var import_item_slots = require("./item-slots.js");
var styles = __toESM(require("./Menu.sty.js"));
var import_useFocusRing = require("../../core/styles/useFocusRing.js");
var import_merge_props = require("../../core/utils/merge-props.js");
var import_Flex = require("../../layouts/flex/Flex.js");
var import_textStyle_css = require("../../styles/textStyle.sty.js");
const SubTrigger = /* @__PURE__ */ (0, import_react.forwardRef)(
  ({
    children,
    disabled,
    "aria-label": ariaLabel,
    "aria-labelledby": ariaLabelledBy,
    "aria-describedby": ariaDescribedBy,
    "aria-details": ariaDetails,
    ...remainingProps
  }, forwardedRef) => {
    const slots = (0, import_item_slots.getSubTriggerSlots)(children);
    const { focusProps, focusClassName } = (0, import_useFocusRing.useFocusRing)({
      disabled
    });
    const { isItemHovered, hoverProps } = (0, import_useIsMenuItemHovered.useIsMenuItemHovered)();
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_react_dropdown_menu.SubTrigger,
      {
        ref: forwardedRef,
        disabled,
        className: (0, import_clsx.default)(
          (0, import_textStyle_css.textStyle)({ textStyle: "base-emphasized" }),
          !isItemHovered && focusClassName,
          styles.item({ disabled })
        ),
        ...(0, import_merge_props.mergeProps)(remainingProps, focusProps, hoverProps),
        "aria-label": ariaLabel,
        "aria-labelledby": ariaLabelledBy,
        "aria-describedby": ariaDescribedBy,
        "aria-details": ariaDetails,
        "aria-haspopup": true,
        role: "menuitem",
        asChild: true,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { justifyContent: "space-between", children: [
          slots.prefix,
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: styles.subChildren, children: slots.freeform }),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ChevronRightSmallIcon, {})
        ] })
      }
    );
  }
);
SubTrigger.displayName = "Menu.SubTrigger";
