// @dynatrace/strato-components/esm/filters/timeframe-selector/utils/formatting.js
import { differenceInMilliseconds, intervalToDuration } from "date-fns";
import {
  getDifference,
  getSmallestCommonUnit
} from "./timeframe-calculations.js";
import { TIMEFRAME_SELECTOR_MESSAGES } from "../constants/timeframeSelectorMessages.js";
function formatTimeframeDuration(from, to, intl) {
  if (from === null || to === null) {
    return "";
  }
  if (from.type === "iso8601" || to.type === "iso8601") {
    const difference2 = intervalToDuration({
      start: from.absoluteDate,
      end: to.absoluteDate
    });
    const diffInMilliseconds = differenceInMilliseconds(to.absoluteDate, from.absoluteDate) % 1e3;
    return intl.formatMessage(TIMEFRAME_SELECTOR_MESSAGES.absoluteDateDuration, {
      years: Math.abs(difference2.years ?? 0),
      months: Math.abs(difference2.months ?? 0),
      days: Math.abs(difference2.days ?? 0),
      hours: Math.abs(difference2.hours ?? 0),
      minutes: Math.abs(difference2.minutes ?? 0),
      seconds: Math.abs(difference2.seconds ?? 0),
      milliseconds: Math.abs(diffInMilliseconds ?? 0)
    }).trim();
  }
  const difference = getDifference(from, to);
  const offsetUnit = getSmallestCommonUnit(from, to);
  const timeExpression = TIMEFRAME_SELECTOR_MESSAGES[`${offsetUnit}-duration`];
  return intl.formatMessage(timeExpression, {
    offset: Math.abs(difference === 0 ? 1 : difference)
  }).trim();
}
export {
  formatTimeframeDuration
};
