// @dynatrace/strato-components/esm/filters/timeframe-selector/utils/presets-helpers.js
import { isArray } from "lodash-es";
import {
  timeframeQueryClient,
  TIMEFRAME_PRESETS_QUERY_KEY
} from "./TimeframePresetsClient.js";
import { devLog } from "../../../core/utils/logging.js";
import { parseValue } from "../../../forms/forms-core/time/timeParser.js";
function isTimeframe(timeframe) {
  if (typeof timeframe !== "object") {
    return false;
  }
  return typeof timeframe.from === "string" && typeof timeframe.to === "string";
}
function isTimeframePreset(timeframePreset) {
  if (typeof timeframePreset !== "object") {
    return false;
  }
  const hasValidId = timeframePreset.id !== void 0 && typeof timeframePreset.id === "string";
  const hasValidTimeframe = timeframePreset.timeframe && isTimeframe(timeframePreset.timeframe);
  return hasValidId && hasValidTimeframe;
}
function isUserTimeframePresets(userTimeframePreset) {
  if (typeof userTimeframePreset !== "object" || userTimeframePreset === null) {
    return false;
  }
  const hasValidRecentPresets = isArray(userTimeframePreset.recentPresets) && userTimeframePreset.recentPresets.every(
    (recentPreset) => isTimeframePreset(recentPreset)
  );
  return hasValidRecentPresets;
}
const RECENT_TIMEFRAME_SERVICE_UNAVAILABLE = "TimeframeSelector recent timeframe service is not available.";
const SAVE_RECENT_TIMEFRAME_ERROR = "TimeframeSelector failed to save recent timeframe.";
const GET_RECENT_TIMEFRAME_ERROR = "TimeframeSelector failed to get recent timeframes.";
async function addRecentTimeframePreset(timeframe) {
  const dtRuntimeTimeframes = globalThis.dtRuntime?.timeframes;
  if (!dtRuntimeTimeframes?.addRecentTimeframePreset) {
    devLog(RECENT_TIMEFRAME_SERVICE_UNAVAILABLE);
    return { recentPresets: [] };
  }
  return dtRuntimeTimeframes.addRecentTimeframePreset({ timeframe }).then((presets) => {
    if (presets && isUserTimeframePresets(presets)) {
      return presets;
    }
    return { recentPresets: [] };
  }).catch(() => {
    devLog(SAVE_RECENT_TIMEFRAME_ERROR);
    return { recentPresets: [] };
  });
}
function saveRecentTimeframe(timeframe) {
  void addRecentTimeframePreset({
    from: timeframe.from.value,
    to: timeframe.to.value
  }).then((presets) => {
    timeframeQueryClient.setQueryData(TIMEFRAME_PRESETS_QUERY_KEY, presets);
    return timeframeQueryClient.refetchQueries({
      queryKey: TIMEFRAME_PRESETS_QUERY_KEY,
      type: "all"
    });
  }).catch(() => {
  });
}
async function getTimeframePresets() {
  const dtRuntimeTimeframes = globalThis.dtRuntime?.timeframes;
  if (!dtRuntimeTimeframes?.getTimeframePresets) {
    devLog(RECENT_TIMEFRAME_SERVICE_UNAVAILABLE);
    return { recentPresets: [] };
  }
  return dtRuntimeTimeframes.getTimeframePresets().then((presets) => {
    if (presets && isUserTimeframePresets(presets)) {
      return presets;
    }
    return { recentPresets: [] };
  }).catch(() => {
    devLog(GET_RECENT_TIMEFRAME_ERROR);
    return { recentPresets: [] };
  });
}
function isPresetSelected(value, details, isRecent, precision) {
  if (isRecent) {
    return false;
  }
  const selectedFrom = details?.from?.value;
  const selectedTo = details?.to?.value;
  const { from: presetFrom, to: presetTo } = parseValue(
    value,
    false,
    precision
  );
  return selectedFrom !== void 0 && selectedTo !== void 0 && selectedFrom === presetFrom?.value && selectedTo === presetTo?.value;
}
export {
  addRecentTimeframePreset,
  getTimeframePresets,
  isPresetSelected,
  isTimeframe,
  isTimeframePreset,
  isUserTimeframePresets,
  saveRecentTimeframe
};
