// @dynatrace/strato-components/esm/filters/timeframe-selector/utils/stepper-helpers.js
import { TZDateMini } from "@date-fns/tz";
import {
  add,
  addDays,
  addHours,
  intervalToDuration,
  startOfHour,
  sub
} from "date-fns";
import {
  _parseTimeExpression,
  parseTimeAsTimeValue
} from "@dynatrace-sdk/units";
import { getTimezone } from "@dynatrace-sdk/user-preferences";
import { formatTimeframeDuration } from "./formatting.js";
import {
  getSmallestCommonUnit,
  getTimeDifferenceInUnit
} from "./timeframe-calculations.js";
import { isBiggerThan, isSmallerThan } from "./timeframe-selection.js";
import {
  getTimeValue,
  removePrecision
} from "../../../forms/forms-core/time/timeParser.js";
import { TIMEFRAME_SELECTOR_MESSAGES } from "../constants/timeframeSelectorMessages.js";
function formatStepperValues(from, to, intl, direction, selectedOverCalendar) {
  const adjustedTo = selectedOverCalendar && to ? getTimeValue(
    startOfHour(
      addHours(new TZDateMini(to.absoluteDate, getTimezone()), 1)
    ).toISOString()
  ) ?? null : to;
  const duration = formatTimeframeDuration(from, adjustedTo, intl);
  const prefix = intl.formatMessage(TIMEFRAME_SELECTOR_MESSAGES.prefix);
  const suffix = intl.formatMessage(TIMEFRAME_SELECTOR_MESSAGES.suffix, {
    suffix: direction === "backward" ? "back" : "forward"
  });
  return `${prefix}${duration}${suffix}`.trim();
}
const adjustDate = (date, difference, direction, selectedOverCalendar) => {
  if (selectedOverCalendar) {
    const daysDifference = Math.round(-difference / (1e3 * 60 * 60 * 24));
    date = addDays(
      date,
      direction === "backward" ? -daysDifference : daysDifference
    );
    return date.toISOString();
  }
  date.setTime(
    direction === "backward" ? date.getTime() - Math.abs(difference) : date.getTime() + Math.abs(difference)
  );
  return date.toISOString();
};
const convertOffsetToExpression = (offsetAmount, diff, offsetUnit) => {
  const totalOffset = diff + offsetAmount;
  if (totalOffset > 0) {
    return `now()+${totalOffset}${offsetUnit}`;
  } else if (totalOffset < 0) {
    return `now()${totalOffset}${offsetUnit}`;
  } else {
    return "now()";
  }
};
const clampToMin = (min, max, from, to, precision) => {
  const relativeNow = TZDateMini.tz(getTimezone()).getTime();
  if (!min) {
    return { from, to };
  }
  const isFromBelowMin = isSmallerThan(from.absoluteDate, min.absoluteDate);
  const isToBelowMin = isSmallerThan(to.absoluteDate, min.absoluteDate);
  const difference = intervalToDuration({
    start: from.absoluteDate,
    end: to.absoluteDate
  });
  if (isFromBelowMin && !isToBelowMin) {
    return {
      from: min,
      to
    };
  }
  const _timezonedMin = new TZDateMini(min.absoluteDate, getTimezone());
  const adjustedTimeISO = new TZDateMini(
    add(_timezonedMin, difference),
    getTimezone()
  ).toISOString();
  let timeValueAdjusted = parseTimeAsTimeValue(
    adjustedTimeISO,
    relativeNow,
    precision
  );
  if (max && isBiggerThan(timeValueAdjusted?.absoluteDate, max.absoluteDate)) {
    timeValueAdjusted = max;
  }
  if (isFromBelowMin && isToBelowMin) {
    return {
      from: min,
      to: timeValueAdjusted ?? min
    };
  }
  if (!isFromBelowMin && isToBelowMin) {
    return {
      from,
      to: timeValueAdjusted ?? from
    };
  }
  return { from, to };
};
const clampToMax = (min, max, from, to, precision) => {
  const relativeNow = TZDateMini.tz(getTimezone()).getTime();
  if (!max) {
    return { from, to };
  }
  const isFromAboveMax = isBiggerThan(from.absoluteDate, max.absoluteDate);
  const isToAboveMax = isBiggerThan(to.absoluteDate, max.absoluteDate);
  const difference = intervalToDuration({
    start: from.absoluteDate,
    end: to.absoluteDate
  });
  if (isFromAboveMax && isToAboveMax) {
    const _timezonedMax = new TZDateMini(max.absoluteDate, getTimezone());
    const adjustedFrom = new TZDateMini(
      sub(_timezonedMax, difference),
      getTimezone()
    ).toISOString();
    if (min && isSmallerThan(adjustedFrom, min.absoluteDate)) {
      return { from: min, to: max };
    }
    return {
      from: parseTimeAsTimeValue(adjustedFrom, relativeNow, precision) ?? max,
      to: max
    };
  }
  if (!isFromAboveMax && isToAboveMax) {
    return {
      from,
      to: max
    };
  }
  if (isFromAboveMax && !isToAboveMax) {
    const _timezonedFrom = new TZDateMini(from.absoluteDate, getTimezone());
    let adjustedTo = new TZDateMini(
      sub(_timezonedFrom, difference),
      getTimezone()
    ).toISOString();
    if (min && isSmallerThan(adjustedTo, min.absoluteDate)) {
      adjustedTo = min.absoluteDate;
    }
    return {
      from,
      to: parseTimeAsTimeValue(adjustedTo, relativeNow, precision) ?? from
    };
  }
  return { from, to };
};
const clampToBoundaries = (min, max, from, to, precision) => {
  const relativeNow = TZDateMini.tz(getTimezone()).getTime();
  const parsedMin = parseTimeAsTimeValue(min, relativeNow, precision);
  const parsedMax = parseTimeAsTimeValue(max, relativeNow, precision);
  let result = { from, to };
  result = clampToMin(parsedMin, parsedMax, result.from, result.to, precision);
  result = clampToMax(parsedMin, parsedMax, result.from, result.to, precision);
  return result;
};
const handleAbsolute = (originalValue, min, max, direction, selectedOverCalendar, precision) => {
  const timezone = getTimezone();
  const from = new TZDateMini(originalValue.from.absoluteDate, timezone);
  const to = new TZDateMini(originalValue.to.absoluteDate, timezone);
  const difference = from.getTime() - to.getTime();
  const adjustedFrom = removePrecision(
    adjustDate(from, difference, direction, selectedOverCalendar),
    precision
  );
  const adjustedTo = removePrecision(
    adjustDate(to, difference, direction, selectedOverCalendar),
    precision
  );
  const newTimeframe = {
    from: {
      type: "iso8601",
      absoluteDate: adjustedFrom,
      value: adjustedFrom
    },
    to: {
      type: "iso8601",
      absoluteDate: adjustedTo,
      value: adjustedTo
    }
  };
  return clampToBoundaries(
    min,
    max,
    newTimeframe.from,
    newTimeframe.to,
    precision
  ) ?? originalValue;
};
const handleExpression = (value, min, max, direction, precision) => {
  const relativeNow = TZDateMini.tz(getTimezone()).getTime();
  const fromDetails = _parseTimeExpression(value.from.value);
  const toDetails = _parseTimeExpression(value.to.value);
  const offsetUnit = getSmallestCommonUnit(value.from, value.to);
  const fromOffset = offsetUnit === fromDetails?.offsetUnit ? fromDetails.offsetAmount : getTimeDifferenceInUnit(fromDetails, relativeNow, offsetUnit);
  const toOffset = offsetUnit === toDetails?.offsetUnit ? toDetails.offsetAmount : getTimeDifferenceInUnit(toDetails, relativeNow, offsetUnit);
  if (fromOffset === void 0 || toOffset === void 0) {
    return void 0;
  }
  const difference = Math.abs(fromOffset - toOffset);
  const newFromValue = convertOffsetToExpression(
    fromOffset,
    difference * -1,
    offsetUnit
  );
  const newToValue = convertOffsetToExpression(
    toOffset,
    difference,
    offsetUnit
  );
  const backwardFromParsedTime = parseTimeAsTimeValue(
    newFromValue,
    relativeNow,
    precision
  );
  const forwardToParsedTime = parseTimeAsTimeValue(
    newToValue,
    relativeNow,
    precision
  );
  if (!backwardFromParsedTime || !forwardToParsedTime) {
    return void 0;
  }
  const currentBackwardFromParsedTime = parseTimeAsTimeValue(
    convertOffsetToExpression(0, fromOffset, offsetUnit),
    relativeNow,
    precision
  );
  const currentForwardToParsedTime = parseTimeAsTimeValue(
    convertOffsetToExpression(0, toOffset, offsetUnit),
    relativeNow,
    precision
  );
  if (!currentBackwardFromParsedTime || !currentForwardToParsedTime) {
    return void 0;
  }
  const newTimeframe = direction === "backward" ? {
    from: backwardFromParsedTime,
    to: currentBackwardFromParsedTime
  } : {
    from: currentForwardToParsedTime,
    to: forwardToParsedTime
  };
  return clampToBoundaries(
    min,
    max,
    newTimeframe.from,
    newTimeframe.to,
    precision
  ) ?? value;
};
function handleRoundedTimeframe(value, min, max, direction, precision) {
  const relativeNow = TZDateMini.tz(getTimezone()).getTime();
  const fromDetails = _parseTimeExpression(value.from.value);
  const toDetails = _parseTimeExpression(value.to.value);
  const offsetUnit = getSmallestCommonUnit(value.from, value.to);
  const fromOffset = (
    // same offsetUnit or now() or now()@d.
    offsetUnit === fromDetails?.offsetUnit || fromDetails?.offsetAmount === 0 ? fromDetails.offsetAmount : getTimeDifferenceInUnit(fromDetails, relativeNow, offsetUnit)
  );
  const toOffset = (
    // same offsetUnit or now() or now()@d.
    offsetUnit === toDetails?.offsetUnit || toDetails?.offsetAmount === 0 ? toDetails.offsetAmount : getTimeDifferenceInUnit(toDetails, relativeNow, offsetUnit)
  );
  let backwardFromValue = "";
  let forwardToValue = "";
  if (!fromDetails || !toDetails) {
    return void 0;
  }
  const roundDownUnit = fromDetails.roundDownUnit || toDetails.roundDownUnit;
  if (fromOffset === 0 && toOffset === 0) {
    backwardFromValue = `-1${offsetUnit}@${roundDownUnit}`;
    forwardToValue = `+1${offsetUnit}@${roundDownUnit}`;
  } else {
    const offsetDifference = Math.abs(fromOffset - toOffset);
    backwardFromValue = value.from.value === "now()" && offsetDifference === 1 ? (
      // Shift backward from now to the rounded down unit (e.g., @d).
      `@${roundDownUnit}`
    ) : (
      // Otherwise, if 'from' and 'to' are roundDownExpressions (e.g., -40d@M -> -20d@M).
      convertOffsetToExpression(
        offsetDifference * -1,
        fromOffset,
        offsetUnit
      ) + `@${roundDownUnit}`
    );
    forwardToValue = convertOffsetToExpression(offsetDifference, toOffset, offsetUnit) + `@${roundDownUnit}`;
  }
  const backwardFromParsedTime = parseTimeAsTimeValue(
    backwardFromValue,
    relativeNow,
    precision
  );
  const forwardToParsedTime = parseTimeAsTimeValue(
    forwardToValue,
    relativeNow,
    precision
  );
  if (!backwardFromParsedTime || !forwardToParsedTime) {
    return void 0;
  }
  const backwardFromCandidate = (
    // if the other unit is smaller than this unit convert it to that unit
    fromDetails?.offsetUnit && offsetUnit !== fromDetails?.offsetUnit ? (
      // Convert the offset to the smaller unit
      convertOffsetToExpression(0, fromOffset, offsetUnit) + `@${roundDownUnit}`
    ) : value.from.value
  );
  const forwardToCandidate = (
    // if the other unit is smaller than this unit convert it to that unit
    toDetails?.offsetUnit && offsetUnit !== toDetails?.offsetUnit ? (
      // Convert the offset to the smaller unit
      convertOffsetToExpression(0, toOffset, offsetUnit) + `@${roundDownUnit}`
    ) : value.to.value
  );
  const currentBackwardFromParsedTime = parseTimeAsTimeValue(
    backwardFromCandidate,
    relativeNow,
    precision
  );
  const currentForwardToParsedTime = parseTimeAsTimeValue(
    forwardToCandidate,
    relativeNow,
    precision
  );
  if (!currentBackwardFromParsedTime || !currentForwardToParsedTime) {
    return void 0;
  }
  const newTimeframe = direction === "backward" ? {
    from: backwardFromParsedTime,
    to: currentBackwardFromParsedTime
  } : {
    from: currentForwardToParsedTime,
    to: forwardToParsedTime
  };
  return clampToBoundaries(
    min,
    max,
    newTimeframe.from,
    newTimeframe.to,
    precision
  ) ?? value;
}
function getTimeShift(value, config) {
  const { direction, min, max, selectedOverCalendar, precision } = config;
  const isISO8601 = (type) => type === "iso8601";
  const isExpression = (type) => type === "expression";
  if (!value?.from || !value?.to) {
    return void 0;
  }
  if (isISO8601(value.from.type) || isISO8601(value.to.type)) {
    return handleAbsolute(
      value,
      min,
      max,
      direction,
      selectedOverCalendar,
      precision
    );
  }
  if (isExpression(value.from.type) && isExpression(value.to.type)) {
    const newFrom = _parseTimeExpression(value.from.value);
    const newTo = _parseTimeExpression(value.to.value);
    if (!(newFrom?.roundDownUnit || newTo?.roundDownUnit)) {
      return handleExpression(value, min, max, direction, precision);
    }
    if (
      // can be a roundedTimeframe or now().
      (newFrom?.roundDownUnit || newFrom?.offsetAmount === 0) && (newTo?.roundDownUnit || newTo?.offsetAmount === 0) && // checks if from and to have different roundDownUnits.
      !(newFrom?.roundDownUnit && newTo?.roundDownUnit && newFrom?.roundDownUnit !== newTo?.roundDownUnit)
    ) {
      return handleRoundedTimeframe(value, min, max, direction, precision);
    }
    return handleAbsolute(value, min, max, direction, false, precision);
  }
  return void 0;
}
export {
  formatStepperValues,
  getTimeShift
};
