// @dynatrace/strato-components/esm/filters/timeframe-selector/utils/timeframe-calculations.js
import { TZDateMini } from "@date-fns/tz";
import {
  differenceInMilliseconds,
  differenceInCalendarMonths,
  differenceInCalendarYears,
  differenceInCalendarDays,
  differenceInHours,
  differenceInMinutes,
  differenceInSeconds,
  differenceInCalendarWeeks,
  differenceInCalendarQuarters
} from "date-fns";
import {
  _getDateFromExpression,
  _parseTimeExpression
} from "@dynatrace-sdk/units";
import { getTimezone } from "@dynatrace-sdk/user-preferences";
const unitOrder = {
  y: 9,
  q: 8,
  M: 7,
  w: 6,
  w0: 6,
  w1: 6,
  w2: 6,
  w3: 6,
  w4: 6,
  w5: 6,
  w6: 6,
  d: 5,
  h: 4,
  m: 3,
  s: 2,
  ms: 1
};
function isSmallestCommonDenominatorUnit(left, right) {
  if (left === void 0) {
    return right ?? "ms";
  }
  if (right === void 0) {
    return left ?? "ms";
  }
  return unitOrder[left] < unitOrder[right] ? left : right;
}
function getSmallestCommonUnit(from, to) {
  const newFrom = _parseTimeExpression(from?.value || "");
  const newTo = _parseTimeExpression(to?.value || "");
  if (newFrom && newTo) {
    const unit = isSmallestCommonDenominatorUnit(
      // Use roundDownUnit only when there is no offsetUnit, such as in '@d -> now'.
      // For cases with both roundDownUnit and offsetUnit, like '+1M@d -> +2M@d',
      // rely solely on the offsetUnit to prevent unwanted conversions like '+1d@d'.
      newFrom.offsetUnit || newTo.offsetUnit || newFrom.roundDownUnit,
      newTo.offsetUnit || newFrom.offsetUnit || newTo.roundDownUnit
    );
    switch (unit) {
      case "w0":
      case "w1":
      case "w2":
      case "w3":
      case "w4":
      case "w5":
      case "w6":
        return "w";
      default:
        return unit;
    }
  }
  return "ms";
}
function getDifference(from, to) {
  const timezone = getTimezone();
  const relativeDate = TZDateMini.tz(timezone).getTime();
  const newFrom = _parseTimeExpression(from?.value || "");
  const newTo = _parseTimeExpression(to?.value || "");
  if (newFrom && newTo) {
    const lowestUnit = getSmallestCommonUnit(from, to);
    return getTimeDifferenceInUnit(newFrom, relativeDate, lowestUnit) - getTimeDifferenceInUnit(newTo, relativeDate, lowestUnit);
  }
  if (from?.absoluteDate && to?.absoluteDate) {
    return Math.abs(
      new TZDateMini(from.absoluteDate, timezone).getTime() - new TZDateMini(to.absoluteDate, timezone).getTime()
    );
  }
  return 0;
}
function getTimeDifferenceInUnit(details, relativeDate, targetUnit) {
  if (!details) {
    return 0;
  }
  const newDate = new TZDateMini(relativeDate, getTimezone());
  const expressionDate = _getDateFromExpression(details, relativeDate);
  if (!expressionDate) {
    return 0;
  }
  switch (targetUnit) {
    case "ms":
      return differenceInMilliseconds(expressionDate, newDate);
    case "s":
      return differenceInSeconds(expressionDate, newDate);
    case "m":
      return differenceInMinutes(expressionDate, newDate);
    case "h":
      return differenceInHours(expressionDate, newDate);
    case "d":
      return differenceInCalendarDays(expressionDate, newDate);
    case "w":
      return differenceInCalendarWeeks(expressionDate, newDate);
    case "M":
      return differenceInCalendarMonths(expressionDate, newDate);
    case "q":
      return differenceInCalendarQuarters(expressionDate, newDate);
    case "y":
      return differenceInCalendarYears(expressionDate, newDate);
    default:
      return differenceInMilliseconds(expressionDate, newDate);
  }
}
export {
  getDifference,
  getSmallestCommonUnit,
  getTimeDifferenceInUnit
};
