// @dynatrace/strato-components/esm/filters/timeframe-selector/utils/timeframe-selection.js
import { TZDateMini } from "@date-fns/tz";
import { parseTimeAsTimeValue } from "@dynatrace-sdk/units";
import { getTimezone } from "@dynatrace-sdk/user-preferences";
import { getFromToValues } from "../../../forms/forms-core/time/timeParser.js";
function dateIsMinMax(date, min, max) {
  if (!min && !max || !date) {
    return false;
  }
  const timezone = getTimezone();
  const relativeNow = TZDateMini.tz(timezone).getTime();
  const isoDate = parseTimeAsTimeValue(date.value, relativeNow)?.absoluteDate;
  const tempDate = isoDate ? new TZDateMini(isoDate, timezone) : null;
  if (!tempDate) {
    return false;
  }
  const minIsoDate = parseTimeAsTimeValue(min, relativeNow)?.absoluteDate;
  const maxIsoDate = parseTimeAsTimeValue(max, relativeNow)?.absoluteDate;
  const minDate = minIsoDate ? new TZDateMini(minIsoDate, timezone) : null;
  const maxDate = maxIsoDate ? new TZDateMini(maxIsoDate, timezone) : null;
  if (minDate && minDate.getTime() === tempDate.getTime()) {
    return true;
  }
  if (maxDate && maxDate.getTime() === tempDate.getTime()) {
    return true;
  }
  return false;
}
function isBiggerThan(candidate, threshold) {
  if (!threshold || !candidate) {
    return false;
  }
  const relativeNow = TZDateMini.tz(getTimezone()).getTime();
  const maxIsoDate = parseTimeAsTimeValue(threshold, relativeNow)?.absoluteDate;
  const valueIsoDate = parseTimeAsTimeValue(
    candidate,
    relativeNow
  )?.absoluteDate;
  if (!maxIsoDate || !valueIsoDate) {
    return false;
  }
  const maxDate = new TZDateMini(maxIsoDate, getTimezone());
  const valueDate = new TZDateMini(valueIsoDate, getTimezone());
  return valueDate.getTime() > maxDate.getTime();
}
function isSmallerThan(candidate, threshold) {
  if (!threshold || !candidate) {
    return false;
  }
  const relativeNow = TZDateMini.tz(getTimezone()).getTime();
  const minIsoDate = parseTimeAsTimeValue(threshold, relativeNow)?.absoluteDate;
  const valueIsoDate = parseTimeAsTimeValue(
    candidate,
    relativeNow
  )?.absoluteDate;
  if (!minIsoDate || !valueIsoDate) {
    return false;
  }
  const minDate = new TZDateMini(minIsoDate, getTimezone());
  const valueDate = new TZDateMini(valueIsoDate, getTimezone());
  return valueDate.getTime() < minDate.getTime();
}
function isOutsideRange(value, min, max) {
  if (!value || !min && !max) {
    return false;
  }
  const { valueFrom, valueTo } = getFromToValues(value);
  if (min && (isSmallerThan(valueFrom, min) || isSmallerThan(valueTo, min))) {
    return true;
  }
  if (max && (isBiggerThan(valueFrom, max) || isBiggerThan(valueTo, max))) {
    return true;
  }
  return false;
}
function isTimeframeWithFromAndTo(date) {
  return typeof date?.from === "object" && typeof date?.to === "object" && date.from?.absoluteDate !== "" && date.to?.absoluteDate !== "";
}
function isTimeframeOrFromToString(date) {
  return typeof date?.from === "string" && typeof date?.to === "string" || isTimeframeWithFromAndTo(date);
}
export {
  dateIsMinMax,
  isBiggerThan,
  isOutsideRange,
  isSmallerThan,
  isTimeframeOrFromToString
};
