// @dynatrace/strato-components/tables/DataTable/features/ColumnFontStyle/ColumnFontStyle.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ColumnFontStyle_exports = {};
__export(ColumnFontStyle_exports, {
  DataTableColumnFontStyle: () => DataTableColumnFontStyle,
  useColumnFontStyle: () => useColumnFontStyle
});
module.exports = __toCommonJS(ColumnFontStyle_exports);
var import_react_table = require("@tanstack/react-table");
var import_useControllableState = require("../../../../core/hooks/useControllableState.js");
var import_useDataTableConfig = require("../../../hooks/useDataTableConfig.js");
const DataTableColumnFontStyle = {
  createColumn(column, table) {
    column.toggleFontStyle = (value) => {
      table.setColumnFontStyle((old) => {
        let newValue = value;
        if (newValue === void 0) {
          if (old[column.id] !== void 0) {
            newValue = old[column.id] === "code" ? "text" : "code";
          } else {
            newValue = table.getVisualVariantOptions().fontStyle === "code" ? "text" : "code";
          }
        }
        return {
          ...old,
          [column.id]: newValue
        };
      });
    };
    column.getFontStyle = () => {
      const fontStyleState = table.getState().columnFontStyle;
      if (fontStyleState?.[column.id]) {
        return fontStyleState?.[column.id];
      }
      return column.columnDef.meta?.fontStyle ?? table.getVisualVariantOptions().fontStyle;
    };
  },
  createTable(table) {
    table.setColumnFontStyle = (updater) => {
      const newValue = (0, import_react_table.functionalUpdate)(
        updater,
        table.getState().columnFontStyle ?? {}
      );
      table.options.onColumnFontStyleChange?.(newValue);
    };
  }
};
function useColumnFontStyle(props, options) {
  const configFontStyle = (0, import_useDataTableConfig.useDataTableConfigElement)("fontStyle");
  const [columnFontStyle, onColumnFontStyleChange] = (0, import_useControllableState.useControllableState)({
    value: props.fontStyle ?? configFontStyle,
    defaultValue: props.defaultFontStyle,
    onChange: props.onFontStyleChange
  });
  options.state = { ...options.state, columnFontStyle };
  options.onColumnFontStyleChange = onColumnFontStyleChange;
}
