// @dynatrace/strato-components/charts/xy-chart/context/XYChartThresholds.context.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var XYChartThresholds_context_exports = {};
__export(XYChartThresholds_context_exports, {
  NoThresholdData: () => NoThresholdData,
  XYChartThresholdsContext: () => XYChartThresholdsContext
});
module.exports = __toCommonJS(XYChartThresholds_context_exports);
var import_react = require("react");
const NoThresholdData = {
  none: true,
  thresholds: [],
  offset: {
    left: 0,
    right: 0
  },
  yAxes: {
    left: /* @__PURE__ */ new Set(),
    right: /* @__PURE__ */ new Set()
  }
};
const XYChartThresholdsContext = (0, import_react.createContext)(NoThresholdData);
