// @dynatrace/strato-components/esm/editors/dql-editor/dql/dql-linter.js
import { memoize } from "lodash-es";
import {
  queryAssistanceClient
} from "@dynatrace-sdk/client-query";
const severities = {
  error: "error",
  warning: "warning",
  info: "info"
};
const toSeverity = (severity) => severities[severity.toLowerCase()] || "info";
const toLinterMessage = ({ notifications }) => (notifications ?? []).map(
  (notification) => notification.syntaxPosition && notification.severity ? {
    from: notification.syntaxPosition.start.index,
    to: notification.syntaxPosition.end.index + 1,
    message: notification.message ?? "",
    severity: toSeverity(notification.severity)
  } : void 0
).filter((x) => x !== void 0);
const transformToValidityInfo = (diagnostics) => diagnostics.map(
  (diagnostic) => ({
    ...diagnostic,
    severity: diagnostic.severity === "hint" ? "info" : diagnostic.severity
  })
);
const rawVerify = async (query) => {
  return await queryAssistanceClient.queryVerify({ body: { query } });
};
const TTL_MS = 3 * 1e3;
const memoizedRawVerify = memoize(rawVerify);
const ttlTimers = /* @__PURE__ */ new Map();
const verifyQuery = async (query) => {
  try {
    if (!query.length) {
      return void 0;
    }
    const result = await memoizedRawVerify(query);
    const pendingTimeout = ttlTimers.get(query);
    if (pendingTimeout) {
      clearTimeout(pendingTimeout);
    }
    const timeout = setTimeout(() => {
      memoizedRawVerify.cache.delete(query);
      ttlTimers.delete(query);
    }, TTL_MS);
    ttlTimers.set(query, timeout);
    return result;
  } catch {
    memoizedRawVerify.cache.delete(query);
    const timeout = ttlTimers.get(query);
    if (timeout) {
      clearTimeout(timeout);
      ttlTimers.delete(query);
    }
  }
  return void 0;
};
const dqlLinter = async (content, onValidationStart, onValidationEnd) => {
  onValidationStart?.();
  const result = await verifyQuery(content);
  const dqlErrors = result ? toLinterMessage(result) : [];
  const dqlEditorValidityInfos = transformToValidityInfo(dqlErrors);
  onValidationEnd?.(dqlEditorValidityInfos);
  return dqlEditorValidityInfos;
};
export {
  dqlLinter,
  verifyQuery
};
