// @dynatrace/strato-components/esm/editors/dql-editor/dql/fuzzy-match.js
const SCORE = {
  startingWithSubstring: 3,
  emptyPattern: 3,
  substring: 2,
  fuzzyMatch: 1,
  noMatch: 0
};
const MAX_FUZZY_SUGGESTION_LENGTH = 5e4;
function findPosOf(substring, str, strTrim = 0) {
  const startIndex = str.indexOf(substring);
  if (startIndex >= 0) {
    return [startIndex + strTrim, startIndex + substring.length + strTrim];
  }
  return void 0;
}
function normalizeString(str) {
  let normalizedStr = str.trimStart();
  const trimStart = str.length - normalizedStr.length;
  normalizedStr = normalizedStr.trimEnd().toLowerCase();
  return [normalizedStr, trimStart];
}
function findSubstringMatch(pattern, str) {
  if (pattern.length === 0) {
    return { match: true, pos: void 0, score: SCORE.emptyPattern };
  }
  const [normalizedPattern] = normalizeString(pattern);
  const [normalizedStr, strTrim] = normalizeString(str);
  const pos = findPosOf(normalizedPattern, normalizedStr, strTrim);
  if (pos !== void 0) {
    const score = pos[0] === strTrim ? SCORE.startingWithSubstring : SCORE.substring;
    return {
      match: true,
      pos,
      score
    };
  }
  return { match: false, score: 0 };
}
function findRangeStart(suggestionText, patternText, rangeEnd) {
  let backtrackSuggestionIndex = rangeEnd;
  let backtrackPatternIndex = patternText.length - 1;
  while (backtrackPatternIndex >= 0) {
    if (suggestionText[backtrackSuggestionIndex] === patternText[backtrackPatternIndex]) {
      backtrackPatternIndex--;
    }
    backtrackSuggestionIndex--;
  }
  return backtrackSuggestionIndex + 1;
}
function findFuzzyMatchRange(patternText, suggestionText) {
  const patternLength = patternText.length;
  const suggestionLength = suggestionText.length;
  if (patternLength > suggestionLength) {
    return void 0;
  }
  if (suggestionLength > MAX_FUZZY_SUGGESTION_LENGTH) {
    return void 0;
  }
  let bestRangeStart = -1;
  let bestRangeEnd = -1;
  let bestRangeLength = Number.POSITIVE_INFINITY;
  let suggestionIndex = 0;
  let patternIndex = 0;
  while (suggestionIndex < suggestionLength) {
    const remainingSuggestionTextLength = suggestionLength - suggestionIndex;
    const remainingPatternTextLength = patternLength - patternIndex;
    if (remainingSuggestionTextLength < remainingPatternTextLength) {
      break;
    }
    if (suggestionText[suggestionIndex] !== patternText[patternIndex]) {
      suggestionIndex++;
      continue;
    }
    patternIndex++;
    suggestionIndex++;
    if (patternIndex === patternLength) {
      const rangeEnd = suggestionIndex;
      const rangeStart = findRangeStart(suggestionText, patternText, rangeEnd);
      const rangeLength = rangeEnd - rangeStart;
      if (rangeLength < bestRangeLength) {
        bestRangeStart = rangeStart;
        bestRangeEnd = rangeEnd;
        bestRangeLength = rangeLength;
      }
      suggestionIndex = rangeStart + 1;
      patternIndex = 0;
    }
  }
  if (bestRangeStart >= 0) {
    return [bestRangeStart, bestRangeEnd];
  }
  return void 0;
}
function findFuzzyMatch(pattern, str) {
  if (pattern.length === 0) {
    return { match: true, pos: void 0, score: SCORE.emptyPattern };
  }
  const [normalizedPattern] = normalizeString(pattern);
  const [normalizedStr, strTrim] = normalizeString(str);
  const range = findFuzzyMatchRange(normalizedPattern, normalizedStr);
  if (range !== void 0) {
    return {
      match: true,
      pos: [range[0] + strTrim, range[1] + strTrim],
      score: SCORE.fuzzyMatch
    };
  }
  return {
    match: false,
    score: SCORE.noMatch
  };
}
function fuzzyMatch(pattern, str) {
  const substringMatch = findSubstringMatch(pattern, str);
  if (substringMatch.match) {
    return substringMatch;
  }
  return findFuzzyMatch(pattern, str);
}
const maxLabelLength = 53;
const middleEllipsisStartPos = 26;
const middleEllipsisLength = 1;
const afterMiddleEllipsisLength = maxLabelLength - middleEllipsisStartPos - middleEllipsisLength;
function getMiddleEllipsisEndPos(label) {
  return label.length - afterMiddleEllipsisLength;
}
function labelShouldBeTruncated(label) {
  return label.length > maxLabelLength;
}
function convertToMiddleEllipsis(label) {
  return `${label.slice(0, middleEllipsisStartPos)}\u2026${label.slice(
    getMiddleEllipsisEndPos(label)
  )}`;
}
function convertPosToMiddleEllipsis(position, label, defaultValue) {
  if (position <= middleEllipsisStartPos) {
    return position;
  }
  const endPos = getMiddleEllipsisEndPos(label);
  if (position > endPos) {
    return middleEllipsisStartPos + middleEllipsisLength + (position - endPos);
  }
  return defaultValue;
}
function getFuzzyMatchFullLabelPosition(pattern, fullLabel) {
  if (fullLabel === void 0) {
    return [];
  }
  const isTruncated = labelShouldBeTruncated(fullLabel);
  const pos = fuzzyMatch(pattern, fullLabel).pos;
  if (pos === void 0) {
    return void 0;
  }
  if (!isTruncated) {
    return pos;
  }
  return [
    convertPosToMiddleEllipsis(pos[0], fullLabel, middleEllipsisStartPos),
    convertPosToMiddleEllipsis(
      pos[1],
      fullLabel,
      middleEllipsisStartPos + middleEllipsisLength
    )
  ];
}
export {
  convertToMiddleEllipsis,
  fuzzyMatch,
  getFuzzyMatchFullLabelPosition,
  labelShouldBeTruncated
};
