// @dynatrace/strato-components/esm/editors/dql-editor/dql/open-autocomplete.js
import { CompletionContext } from "@codemirror/autocomplete";
import { verifyQuery } from "./dql-linter.js";
function isStartOfCommandOrParameter(context) {
  return context.matchBefore(/[,|]\s+$/) !== null;
}
function isTypingCommandPrefixAtCommandBoundary(context) {
  return context.matchBefore(/(?:^|[|,]\s*|\*\/\s*)[A-Za-z][A-Za-z0-9]*$/) !== null;
}
function isTypingParameterPrefixAfterCommand(context) {
  const commandAndParameter = context.matchBefore(
    /^[A-Za-z][A-Za-z0-9]*\s+[A-Za-z_][A-Za-z0-9_.]*\s*$/
  );
  return commandAndParameter !== null;
}
function isBeforeCommentOrEndOfContent(state) {
  const afterCursor = state.doc.slice(state.selection.main.to).toString();
  const afterCursorTrimmed = afterCursor.trimStart();
  return afterCursorTrimmed === "" || afterCursorTrimmed.startsWith("//") || afterCursorTrimmed.startsWith("/*");
}
async function isInvalidAtEnd(query) {
  const trimmedQuery = query.trimEnd();
  const { valid = true, notifications = [] } = await verifyQuery(trimmedQuery) ?? {};
  const hasErrorNotificationAtEnd = notifications?.some(
    (notification) => notification.severity === "ERROR" && notification.syntaxPosition?.end.index === trimmedQuery.length - 1
  );
  return !valid && hasErrorNotificationAtEnd;
}
async function openAutocomplete(update, state) {
  const { isTyping, isDeleting, isFocused } = update;
  const context = new CompletionContext(state, state.selection.main.to, false);
  if (isTyping && (isTypingCommandPrefixAtCommandBoundary(context) || isTypingParameterPrefixAfterCommand(context)) && isBeforeCommentOrEndOfContent(state)) {
    return true;
  }
  if (isFocused && state.doc.toString().trim() === "") {
    return true;
  }
  const hasWhitespaceBefore = context.matchBefore(/\s/);
  if ((isTyping || isDeleting) && context.tokenBefore(["keyword"]) && hasWhitespaceBefore) {
    return true;
  }
  if (isTyping && isStartOfCommandOrParameter(context)) {
    return true;
  }
  if (isTyping && hasWhitespaceBefore) {
    const query = state.doc.slice(0, state.selection.main.to).toString();
    if (await isInvalidAtEnd(query)) {
      return true;
    }
  }
  const characterAfter = state.doc.slice(state.selection.main.to, state.selection.main.to + 1).toString();
  const hasSpaceOrEndOfContentAfter = characterAfter.trim().length === 0;
  if (isDeleting && hasSpaceOrEndOfContentAfter) {
    return true;
  }
  return false;
}
export {
  isInvalidAtEnd,
  openAutocomplete
};
