// @dynatrace/strato-components/esm/editors/dql-editor/dql/SuggestionPart.js
import { jsx, jsxs } from "react/jsx-runtime";
import { DocumentationLink } from "./DocumentationLink.js";
import { suggestionPartSectionInfoCSS } from "./SuggestionPart.sty.js";
import { CodeSnippet } from "../../../content/code-snippet/CodeSnippet.js";
import { Flex } from "../../../layouts/flex/Flex.js";
import { globalSprinkles } from "../../../styles/safe-sprinkles.js";
import { Text } from "../../../typography/text/Text.js";
const SuggestionPart = (props) => {
  const { suggestionPart } = props;
  return /* @__PURE__ */ jsxs(Flex, { flexDirection: "column", gap: 0, children: [
    /* @__PURE__ */ jsxs(Text, { textStyle: "base", className: suggestionPartSectionInfoCSS, children: [
      suggestionPart.info,
      " ",
      /* @__PURE__ */ jsx(DocumentationLink, { suggestionPart })
    ] }),
    suggestionPart.synopsis && /* @__PURE__ */ jsx(
      CodeSnippet,
      {
        language: "dql",
        showLineNumbers: false,
        showCopyAction: false,
        lineBreaks: true,
        className: globalSprinkles({ marginTop: 4 }),
        children: suggestionPart.synopsis
      }
    )
  ] });
};
export {
  SuggestionPart
};
