// @dynatrace/strato-components/notifications/notification-settings/NotificationSettings.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var NotificationSettings_exports = {};
__export(NotificationSettings_exports, {
  NotificationProvider: () => NotificationProvider,
  NotificationSettings: () => NotificationSettings
});
module.exports = __toCommonJS(NotificationSettings_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_Content = require("./components/Content.js");
var import_EventNotificationClient = require("./components/EventNotificationClient.js");
var import_PermissionInfo = require("./components/PermissionInfo.js");
var import_ResourceNotificationClient = require("./components/ResourceNotificationClient.js");
var import_ToggleButton = require("./components/ToggleButton.js");
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useSafeForwardProps = require("../../core/hooks/useSafeForwardProps.js");
var import_Flex = require("../../layouts/flex/Flex.js");
const COMPONENT_NAME = "NotificationSettings";
function NotificationProvider(props) {
  if ((0, import_ResourceNotificationClient.isResourceNotificationConfig)(props.type, props.config)) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ResourceNotificationClient.ResourceNotificationClient, { config: props.config, children: props.children });
  }
  if ((0, import_EventNotificationClient.isEventNotificationConfig)(props.type, props.config)) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_EventNotificationClient.EventNotificationClient, { config: props.config, children: props.children });
  }
  return props.children;
}
const NotificationSettings = (0, import_react.forwardRef)((props, ref) => {
  const {
    config,
    children,
    id,
    type,
    style: consumerStyle,
    className: consumerClassName,
    "data-testid": dataTestId,
    "data-dtrum-mask": dtRumMask,
    "data-dtrum-allow": dtRumAllow,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const ariaLabelingProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(restProps);
  const forwardProps = (0, import_useSafeForwardProps.useSafeForwardProps)(restProps, ariaLabelingProps);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(NotificationProvider, { type, config, children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_Flex.Flex,
    {
      id,
      ref,
      flexDirection: "column",
      gap: 12,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dtRumMask,
      "data-dtrum-allow": dtRumAllow,
      ...ariaLabelingProps,
      ...behavioralTrackingProps,
      ...forwardProps,
      "data-dt-component": COMPONENT_NAME,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Content.Content, { children: props.children }),
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { alignItems: "center", children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ToggleButton.NotificationToggleButton, {}),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_PermissionInfo.NotificationPermissionInfo, {})
        ] })
      ]
    }
  ) });
});
NotificationSettings.displayName = COMPONENT_NAME;
