// @dynatrace/strato-components/esm/layouts/page-layout/shared/portal-slots.js
import { jsx } from "react/jsx-runtime";
import {
  forwardRef,
  useMemo,
  useRef,
  useSyncExternalStore
} from "react";
import { createPortal } from "react-dom";
import { usePageLayoutPortalContext } from "./PageLayoutPortalContext.js";
import {
  contentDefaults,
  detailsDefaults,
  sidebarDefaults
} from "./slot-defaults.js";
import { useRegisteredSlot } from "./useRegisteredSlot.js";
import { DetailsControlBar } from "../details-layout/DetailsControlBar.js";
import { DetailsLayoutContext } from "../details-layout/DetailsLayoutContext.js";
const toRef = (ref) => ref ?? void 0;
function useShallowStable(value) {
  const ref = useRef(value);
  const prev = ref.current;
  const prevKeys = Object.keys(prev);
  const nextKeys = Object.keys(value);
  let equal = prevKeys.length === nextKeys.length;
  if (equal) {
    for (const key of nextKeys) {
      if (!Object.is(prev[key], value[key])) {
        equal = false;
        break;
      }
    }
  }
  if (!equal) {
    ref.current = value;
  }
  return ref.current;
}
const PageLayoutHeader = forwardRef(
  function PageLayoutHeader2({ children, ...domProps }, ref) {
    const { hosts } = usePageLayoutPortalContext();
    const stableDomProps = useShallowStable(domProps);
    const refValue = toRef(ref);
    const config = useMemo(
      () => ({ domProps: stableDomProps, ref: refValue }),
      [stableDomProps, refValue]
    );
    useRegisteredSlot("header", config);
    return createPortal(children, hosts.header);
  }
);
const PageLayoutSidebar = forwardRef(
  function PageLayoutSidebar2({
    children,
    defaultWidth = sidebarDefaults.defaultWidth,
    minWidth = sidebarDefaults.minWidth,
    maxWidth = sidebarDefaults.maxWidth,
    resizable = sidebarDefaults.resizable,
    collapsed,
    defaultCollapsed = sidebarDefaults.defaultCollapsed,
    onCollapsedChange,
    breakpoint = sidebarDefaults.breakpoint,
    onResize,
    ...domProps
  }, ref) {
    const { hosts } = usePageLayoutPortalContext();
    const stableDomProps = useShallowStable(domProps);
    const refValue = toRef(ref);
    const config = useMemo(
      () => ({
        config: {
          defaultWidth,
          minWidth,
          maxWidth,
          resizable,
          collapsed,
          defaultCollapsed,
          onCollapsedChange,
          breakpoint,
          onResize
        },
        domProps: stableDomProps,
        ref: refValue
      }),
      [
        defaultWidth,
        minWidth,
        maxWidth,
        resizable,
        collapsed,
        defaultCollapsed,
        onCollapsedChange,
        breakpoint,
        onResize,
        stableDomProps,
        refValue
      ]
    );
    useRegisteredSlot("sidebar", config);
    return createPortal(children, hosts.sidebar);
  }
);
const PageLayoutContent = forwardRef(
  function PageLayoutContent2({
    children,
    minWidth = contentDefaults.minWidth,
    ...domProps
  }, ref) {
    const { hosts } = usePageLayoutPortalContext();
    const stableDomProps = useShallowStable(domProps);
    const refValue = toRef(ref);
    const config = useMemo(
      () => ({
        config: { minWidth },
        domProps: stableDomProps,
        ref: refValue
      }),
      [minWidth, stableDomProps, refValue]
    );
    useRegisteredSlot("content", config);
    return createPortal(children, hosts.content);
  }
);
const PageLayoutDetails = forwardRef(
  function PageLayoutDetails2({
    children,
    minWidth = detailsDefaults.minWidth,
    maxWidth = detailsDefaults.maxWidth,
    defaultWidth = detailsDefaults.defaultWidth,
    resizable = detailsDefaults.resizable,
    collapsed,
    defaultCollapsed = detailsDefaults.defaultCollapsed,
    onCollapsedChange,
    layout,
    defaultLayout = detailsDefaults.defaultLayout,
    onLayoutChange,
    breakpoint = detailsDefaults.breakpoint,
    onResize,
    ...domProps
  }, ref) {
    const { hosts, registry } = usePageLayoutPortalContext();
    const stableDomProps = useShallowStable(domProps);
    const refValue = toRef(ref);
    const config = useMemo(
      () => ({
        config: {
          minWidth,
          maxWidth,
          defaultWidth,
          resizable,
          collapsed,
          defaultCollapsed,
          onCollapsedChange,
          layout,
          defaultLayout,
          onLayoutChange,
          breakpoint,
          onResize
        },
        domProps: stableDomProps,
        ref: refValue
      }),
      [
        minWidth,
        maxWidth,
        defaultWidth,
        resizable,
        collapsed,
        defaultCollapsed,
        onCollapsedChange,
        layout,
        defaultLayout,
        onLayoutChange,
        breakpoint,
        onResize,
        stableDomProps,
        refValue
      ]
    );
    useRegisteredSlot("details", config);
    const detailsLayoutContext = useSyncExternalStore(
      registry.subscribe,
      () => registry.getSnapshot().detailsLayoutContext
    );
    return createPortal(
      /* @__PURE__ */ jsx(DetailsLayoutContext.Provider, { value: detailsLayoutContext, children }),
      hosts.details
    );
  }
);
const PageLayoutDetailsWithControlBar = Object.assign(
  PageLayoutDetails,
  { ControlBar: DetailsControlBar }
);
export {
  PageLayoutContent,
  PageLayoutDetails,
  PageLayoutDetailsWithControlBar,
  PageLayoutHeader,
  PageLayoutSidebar
};
