// @dynatrace/strato-components/esm/layouts/page-layout/shared/useBreakpointCollapse.js
import {
  useRef,
  useState
} from "react";
import useResizeObserver from "use-resize-observer";
import { useControllableState } from "../../../core/hooks/useControllableState.js";
function useBreakpointCollapse({
  collapsed: collapsedProps,
  breakpoint,
  containerRef
}) {
  const [containerWidth, setContainerWidth] = useState(0);
  const [collapsed, setCollapsed] = useControllableState({
    value: collapsedProps.value,
    defaultValue: collapsedProps.defaultValue,
    onChange: collapsedProps.onChange
  });
  const prevBelowBreakpointRef = useRef(null);
  useResizeObserver({
    ref: containerRef,
    onResize: (size) => {
      if (!size.width) {
        return;
      }
      setContainerWidth(size.width);
      const isBelowBreakpoint2 = size.width < breakpoint;
      const prevBelowBreakpoint = prevBelowBreakpointRef.current;
      prevBelowBreakpointRef.current = isBelowBreakpoint2;
      if (prevBelowBreakpoint === isBelowBreakpoint2) {
        return;
      }
      if (prevBelowBreakpoint === null && collapsed === isBelowBreakpoint2) {
        return;
      }
      setCollapsed(isBelowBreakpoint2);
    }
  });
  const isBelowBreakpoint = containerWidth !== 0 && containerWidth < breakpoint;
  return { containerWidth, collapsed, setCollapsed, isBelowBreakpoint };
}
export {
  useBreakpointCollapse
};
