// @dynatrace/strato-components/charts/core/components/interactions-overlay/constants.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var constants_exports = {};
__export(constants_exports, {
  KeyboardControls: () => KeyboardControls,
  LARGE_STEP_FACTOR: () => LARGE_STEP_FACTOR,
  MIN_ZOOM_STEP: () => MIN_ZOOM_STEP,
  SELECTION_EDGE_THRESHOLD: () => SELECTION_EDGE_THRESHOLD,
  SELECTION_RIGHT_OVERFLOW: () => SELECTION_RIGHT_OVERFLOW,
  SMALL_STEP_FACTOR: () => SMALL_STEP_FACTOR,
  ZOOM_FACTOR: () => ZOOM_FACTOR
});
module.exports = __toCommonJS(constants_exports);
var KeyboardControls = /* @__PURE__ */ ((KeyboardControls2) => {
  KeyboardControls2["LEFT_ARROW"] = "ArrowLeft";
  KeyboardControls2["RIGHT_ARROW"] = "ArrowRight";
  KeyboardControls2["UP_ARROW"] = "ArrowUp";
  KeyboardControls2["DOWN_ARROW"] = "ArrowDown";
  KeyboardControls2["ENTER"] = "Enter";
  KeyboardControls2["ESCAPE"] = "Escape";
  KeyboardControls2["CMD"] = "Meta";
  KeyboardControls2["CONTROL"] = "Control";
  KeyboardControls2["ALT"] = "Alt";
  KeyboardControls2["SPACE"] = " ";
  KeyboardControls2["KEYBOARD_EXPLORE_MODE_SHORTCUT"] = "e";
  KeyboardControls2["KEYBOARD_EXPLORE_MODE_SHORTCUT_ALTERNATIVE"] = "E";
  KeyboardControls2["KEYBOARD_PAN_MODE_SHORTCUT"] = "p";
  KeyboardControls2["KEYBOARD_PAN_MODE_SHORTCUT_ALTERNATIVE"] = "P";
  KeyboardControls2["KEYBOARD_ZOOM_MODE_SHORTCUT"] = "z";
  KeyboardControls2["KEYBOARD_ZOOM_MODE_SHORTCUT_ALTERNATIVE"] = "Z";
  KeyboardControls2["KEYBOARD_RESET_SHORTCUT"] = "r";
  KeyboardControls2["KEYBOARD_RESET_SHORTCUT_ALTERNATIVE"] = "R";
  return KeyboardControls2;
})(KeyboardControls || {});
const MIN_ZOOM_STEP = 10;
const ZOOM_FACTOR = 0.1;
const SMALL_STEP_FACTOR = 0.01;
const LARGE_STEP_FACTOR = 0.02;
const SELECTION_EDGE_THRESHOLD = 2;
const SELECTION_RIGHT_OVERFLOW = 1;
