// @dynatrace/strato-components/charts/core/components/interactions-overlay/InteractionsOverlay.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var InteractionsOverlay_exports = {};
__export(InteractionsOverlay_exports, {
  InteractionsOverlay: () => InteractionsOverlay
});
module.exports = __toCommonJS(InteractionsOverlay_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_InteractionsOverlay_css = require("./InteractionsOverlay.sty.js");
var import_get_active_mode = require("./utils/get-active-mode.js");
var import_get_explore_interations = require("./utils/get-explore-interations.js");
var import_get_gestures = require("./utils/get-gestures.js");
var import_get_interface_interactions = require("./utils/get-interface-interactions.js");
var import_get_pan_interactions = require("./utils/get-pan-interactions.js");
var import_get_zoom_interactions = require("./utils/get-zoom-interactions.js");
var import_handle_keydown = require("./utils/handle-keydown.js");
var import_is_own_event = require("./utils/is-own-event.js");
var import_useMergeRefs = require("../../../../core/hooks/useMergeRefs.js");
var import_usePointerEvents = require("../../hooks/usePointerEvents.js");
var import_usePreventScrollOnWheel = require("../../hooks/usePreventScrollOnWheel.js");
const InteractionsOverlay = (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    style,
    className,
    mode,
    pinned,
    selectionArea,
    width,
    height,
    onModeChange,
    onFocus,
    onHover,
    onLeave,
    onPanStart,
    onPanChange,
    onPanEnd,
    onTogglePin,
    onReset,
    onSelectStart,
    onSelectChange,
    onSelectEnd,
    onZoomStart,
    onZoomChange,
    onZoomEnd,
    isZoomDisabled,
    isPanDisabled,
    isZoomXDisabled,
    children,
    ...restProps
  } = props;
  const zoomDisabled = selectionArea ? true : props.isZoomDisabled;
  const panDisabled = selectionArea ? true : props.isPanDisabled;
  const panDeltaRef = (0, import_react.useRef)(void 0);
  const lastMousePositionRef = (0, import_react.useRef)(void 0);
  const isPointerDownRef = (0, import_react.useRef)(false);
  const eventCancelled = (0, import_react.useRef)(false);
  const handlerRef = (0, import_react.useRef)(null);
  const [cursorMode, setCursorMode] = (0, import_react.useState)(
    mode
  );
  const htmlRef = (0, import_react.useRef)(null);
  (0, import_usePreventScrollOnWheel.usePreventScrollOnWheel)(htmlRef);
  const rootRef = (0, import_useMergeRefs.useMergeRefs)([htmlRef, forwardedRef]);
  const overlappingInteractionsProps = {
    selectionArea,
    eventCancelled,
    onSelectEnd,
    lastMousePositionRef,
    pinned,
    onHover,
    onTogglePin,
    setCursorMode,
    height,
    onReset,
    width,
    onZoomChange,
    onPanChange,
    panDeltaRef,
    onPanStart,
    onPanEnd
  };
  const handleMouseLeave = () => {
    onLeave?.({ triggeredBy: "pointer" });
    isPointerDownRef.current = false;
    lastMousePositionRef.current = void 0;
  };
  const handleFocus = (event) => {
    if (handlerRef.current) {
      handlerRef.current();
      handlerRef.current = null;
    }
    onFocus?.(event);
  };
  const handleBlur = (event) => {
    const nextFocusElement = event.relatedTarget;
    if (!event.target?.contains(nextFocusElement)) {
      onLeave?.({ triggeredBy: "keyboard" });
    }
  };
  const handleWheel = (event) => {
    interactions[(0, import_get_active_mode.getActiveMode)(mode, event, zoomDisabled)].onWheel(event);
  };
  const handleKeyboard = (event) => (0, import_handle_keydown.handleKeyDown)({
    event,
    lastMousePositionRef,
    panDeltaRef,
    selectionArea,
    eventCancelled,
    onSelectEnd,
    onModeChange,
    interaction: interactions[(0, import_get_active_mode.getActiveMode)(mode, event, zoomDisabled, isPanDisabled)]
  });
  const interactions = {
    explore: (0, import_get_explore_interations.getExploreInteractions)({
      ...overlappingInteractionsProps,
      onModeChange,
      onSelectChange,
      onSelectStart,
      isZoomXDisabled
    }),
    zoom: (0, import_get_zoom_interactions.getZoomInteractions)({
      ...overlappingInteractionsProps,
      cursorMode,
      onSelectChange,
      onSelectStart
    }),
    pan: (0, import_get_pan_interactions.getPanInteractions)({
      ...overlappingInteractionsProps,
      onModeChange,
      cursorMode
    })
  };
  const {
    handleMouseDown,
    handleMouseMove,
    handleMouseUp,
    handleTouchStart,
    handleTouchMove,
    handleTouchEnd
  } = (0, import_usePointerEvents.usePointerEvents)({
    ...(0, import_get_interface_interactions.getInterfaceInteractions)(
      interactions,
      mode,
      zoomDisabled,
      isZoomXDisabled,
      panDisabled
    ),
    ...(0, import_get_gestures.getGestures)({
      lastMousePositionRef,
      panDeltaRef,
      onZoomStart,
      onZoomChange,
      onZoomEnd,
      onPanStart,
      onPanChange,
      onPanEnd
    }),
    eventCancelled
  });
  const handleDocumentMouseUp = (e) => {
    const { clientX, clientY } = e;
    if (htmlRef?.current) {
      handleMouseUp({
        ...e,
        currentTarget: htmlRef.current,
        clientX,
        clientY
      });
      isPointerDownRef.current = false;
      lastMousePositionRef.current = void 0;
    }
  };
  const handleDocumentMouseMove = (e) => {
    const { clientX, clientY } = e;
    if (htmlRef?.current) {
      handleMouseMove({
        ...e,
        currentTarget: htmlRef.current,
        clientX,
        clientY
      });
    }
  };
  const registerDocumentListeners = () => {
    if (!handlerRef.current) {
      document.addEventListener("mouseup", handleDocumentMouseUp);
      document.addEventListener("pointermove", handleDocumentMouseMove);
      handlerRef.current = () => {
        document.removeEventListener("mouseup", handleDocumentMouseUp);
        document.removeEventListener("pointermove", handleDocumentMouseMove);
      };
    }
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      ref: rootRef,
      role: "presentation",
      tabIndex: 0,
      className: (0, import_clsx.default)(
        (0, import_InteractionsOverlay_css.chartEventsWrapper)({ mode: cursorMode }),
        className
      ),
      style,
      onMouseDown: (event) => {
        if ((0, import_is_own_event.isOwnEvent)(event)) {
          isPointerDownRef.current = true;
          handleMouseDown(event);
        }
      },
      onMouseMove: (event) => {
        if ((0, import_is_own_event.isOwnEvent)(event)) {
          handleMouseMove(event);
        }
      },
      onMouseUp: (event) => {
        if ((0, import_is_own_event.isOwnEvent)(event)) {
          handleMouseUp(event);
          isPointerDownRef.current = false;
        }
      },
      onMouseLeave: (e) => {
        if (isPointerDownRef.current && lastMousePositionRef.current && htmlRef.current) {
          registerDocumentListeners();
        } else {
          handleMouseLeave();
        }
      },
      onTouchStart: (event) => {
        if ((0, import_is_own_event.isOwnEvent)(event)) {
          handleTouchStart(event);
        }
      },
      onTouchMove: (event) => {
        if ((0, import_is_own_event.isOwnEvent)(event)) {
          handleTouchMove(event);
        }
      },
      onTouchEnd: (event) => {
        if ((0, import_is_own_event.isOwnEvent)(event)) {
          handleTouchEnd(event);
        }
      },
      onMouseEnter: () => handleFocus(),
      onFocus: handleFocus,
      onBlur: handleBlur,
      onKeyDown: handleKeyboard,
      onWheel: handleWheel,
      ...restProps,
      children
    }
  );
});
InteractionsOverlay["displayName"] = "InteractionsOverlay";
