// @dynatrace/strato-geo/esm/map/providers/imperative-handler.provider.js
import { useMap } from "@vis.gl/react-maplibre";
import {
  useImperativeHandle
} from "react";
import { FIT_BOUNDS_OPTIONS } from "../constants.js";
import { useDataBoundingBox } from "../hooks/use-data-bounding-box.js";
import { useInitialViewContext } from "../hooks/use-initial-view-context.js";
import { useMapRawData } from "../hooks/use-map-raw-data.js";
import { useSetStateZoom } from "../store/store.js";
import { buildAndDownloadCsv } from "../utils/build-and-download-csv.js";
const MapViewImperativeHandler = ({
  forwardedRef,
  containerRef,
  children
}) => {
  const layersData = useMapRawData();
  const { current: map } = useMap();
  const { longitude = 0, latitude = 0, zoom = 0 } = useInitialViewContext();
  const boundingBox = useDataBoundingBox();
  const setZoomState = useSetStateZoom();
  const zoomOutEnabled = map && map?.getMinZoom() !== map?.getZoom();
  const zoomInEnabled = map && map?.getMaxZoom() !== map?.getZoom();
  const handleZoomToFit = () => {
    map?.fitBounds(boundingBox, FIT_BOUNDS_OPTIONS);
    setZoomState({ zoomToFitEnabled: false, zoomToFitInProgress: true });
  };
  useImperativeHandle(forwardedRef, () => ({
    element: containerRef.current,
    downloadData: () => {
      buildAndDownloadCsv(layersData);
    },
    zoomIn: () => zoomInEnabled && map?.zoomIn(),
    zoomOut: () => zoomOutEnabled && map?.zoomOut(),
    zoomToFit: () => handleZoomToFit(),
    reset: () => map?.flyTo({ center: [longitude, latitude], zoom })
  }));
  return children;
};
export {
  MapViewImperativeHandler
};
