// @dynatrace/strato-components/charts/core/components/annotations/utils/get-marker-brightness-based-on-state.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var get_marker_brightness_based_on_state_exports = {};
__export(get_marker_brightness_based_on_state_exports, {
  getMarkerBrightnessBasedOnState: () => getMarkerBrightnessBasedOnState
});
module.exports = __toCommonJS(get_marker_brightness_based_on_state_exports);
const MARKER_DEFAULT_BRIGHTNESS = 1;
const mapStateBrightness = {
  dark: {
    default: MARKER_DEFAULT_BRIGHTNESS,
    hovered: 1.3,
    active: 1.5,
    inactive: 0.65
  },
  light: {
    default: MARKER_DEFAULT_BRIGHTNESS,
    hovered: 0.8,
    active: 0.65,
    inactive: 1.3
  }
};
const getMarkerBrightnessBasedOnState = (theme, state) => mapStateBrightness[theme][state];
