// @dynatrace/strato-components/charts/gantt-chart/hooks/use-set-tooltip-state.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var use_set_tooltip_state_exports = {};
__export(use_set_tooltip_state_exports, {
  useSetTooltipState: () => useSetTooltipState,
  useTooltipContent: () => useTooltipContent,
  useTooltipPinnedStatus: () => useTooltipPinnedStatus,
  useTooltipPosition: () => useTooltipPosition,
  useTooltipVisibility: () => useTooltipVisibility
});
module.exports = __toCommonJS(use_set_tooltip_state_exports);
var import_react = require("react");
var import_selector = require("../../../core/store/selector.js");
var import_GanttChartStateContext = require("../context/GanttChartStateContext.js");
const useSetTooltipState = () => {
  const { setState } = (0, import_react.useContext)(import_GanttChartStateContext.GanttChartStateContext);
  const setTooltipStateContent = (0, import_react.useCallback)(
    (content) => {
      setState((prev) => {
        return {
          ...prev,
          tooltip: {
            ...prev.tooltip,
            content
          }
        };
      });
    },
    [setState]
  );
  const setTooltipStatePosition = (0, import_react.useCallback)(
    (position) => {
      setState((prev) => {
        return {
          ...prev,
          tooltip: {
            ...prev.tooltip,
            position
          }
        };
      });
    },
    [setState]
  );
  const setTooltipStateVisibility = (0, import_react.useCallback)(
    (visible) => {
      setState((prev) => {
        return {
          ...prev,
          tooltip: {
            ...prev.tooltip,
            visible
          }
        };
      });
    },
    [setState]
  );
  const setTooltipStatePinned = (0, import_react.useCallback)(
    (pinned) => {
      setState((prev) => {
        return {
          ...prev,
          tooltip: {
            ...prev.tooltip,
            pinned
          }
        };
      });
    },
    [setState]
  );
  return {
    setTooltipStateContent,
    setTooltipStatePosition,
    setTooltipStateVisibility,
    setTooltipStatePinned
  };
};
const useSelector = (selector) => (0, import_selector.useStoreSelector)(import_GanttChartStateContext.GanttChartStateContext, selector);
const useTooltipContent = () => useSelector((state) => state.tooltip.content);
const useTooltipPinnedStatus = () => useSelector((state) => state.tooltip.pinned);
const useTooltipPosition = () => useSelector((state) => state.tooltip.position);
const useTooltipVisibility = () => useSelector((state) => state.tooltip.visible);
