// @dynatrace/strato-components/charts/gantt-chart/hooks/use-gantt-entry-handlers.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var use_gantt_entry_handlers_exports = {};
__export(use_gantt_entry_handlers_exports, {
  useGanttEntryHandlers: () => useGanttEntryHandlers
});
module.exports = __toCommonJS(use_gantt_entry_handlers_exports);
var import_use_set_tooltip_state = require("./use-set-tooltip-state.js");
function useGanttEntryHandlers(content) {
  const isTooltipPinned = (0, import_use_set_tooltip_state.useTooltipPinnedStatus)();
  const {
    setTooltipStateContent,
    setTooltipStatePinned,
    setTooltipStateVisibility,
    setTooltipStatePosition
  } = (0, import_use_set_tooltip_state.useSetTooltipState)();
  if (content === void 0) {
    return {
      onMouseEnterHandler: void 0,
      onMouseClickHandler: void 0,
      onMouseLeaveHandler: void 0
    };
  }
  const onMouseEnterHandler = () => {
    setTooltipStateVisibility(true);
    setTooltipStateContent({
      segment: content.segment,
      row: content.row,
      parent: content.parent
    });
  };
  const onMouseMoveHandler = (event) => {
    const { clientX, clientY } = event;
    const position = {
      x: clientX,
      y: clientY
    };
    setTooltipStatePosition(position);
  };
  const onMouseClickHandler = () => {
    setTooltipStatePinned(false);
    setTooltipStateContent({
      segment: content.segment,
      row: content.row,
      parent: content.parent
    });
    requestAnimationFrame(() => setTooltipStatePinned(true));
  };
  const onMouseLeaveHandler = () => {
    !isTooltipPinned && setTooltipStateVisibility(false);
    setTooltipStateContent(void 0);
  };
  return {
    onMouseEnterHandler,
    onMouseClickHandler,
    onMouseLeaveHandler,
    onMouseMoveHandler
  };
}
