// @dynatrace/strato-components/layouts/page-layout/details-layout/DetailsLayout.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DetailsLayout_exports = {};
__export(DetailsLayout_exports, {
  DetailsLayout: () => DetailsLayout
});
module.exports = __toCommonJS(DetailsLayout_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = require("clsx");
var import_react = require("react");
var import_react_intl = require("react-intl");
var styles = __toESM(require("./DetailsLayout.sty.js"));
var import_DetailsLayoutContext = require("./DetailsLayoutContext.js");
var import_Overlay = require("./Overlay.js");
var import_useLayoutState = require("./useLayoutState.js");
var import_useMergeRefs = require("../../../core/hooks/useMergeRefs.js");
var import_useSafeLayoutEffect = require("../../../core/hooks/useSafeLayoutEffect.js");
var import_Drawer = require("../../../overlays/drawer/Drawer.js");
var import_SplitLayout = require("../../split-layout/SplitLayout.js");
var import_width_utils = require("../../split-layout/width-utils.js");
var import_PageLayoutPortalContext = require("../shared/PageLayoutPortalContext.js");
var import_useTeleport = require("../shared/useTeleport.js");
function DetailsLayoutInner({
  details,
  content,
  containerRef,
  persistedPanelWidthRef,
  contentTeleport,
  panelTeleport
}) {
  const intl = (0, import_react_intl.useIntl)();
  const layout = (0, import_useLayoutState.useLayoutState)({ details, containerRef });
  const detailsAriaLabel = intl.formatMessage({
    defaultMessage: "Details",
    id: "o3A99RFWS+N7bJe8",
    description: "Accessible label for the details panel."
  });
  const contextValue = (0, import_react.useMemo)(
    () => ({
      collapsed: layout.collapsed,
      setCollapsed: layout.setCollapsed,
      preferredLayout: layout.preferredLayoutMode,
      setPreferredLayout: layout.setPreferredLayoutMode,
      mode: layout.mode
    }),
    [
      layout.collapsed,
      layout.setCollapsed,
      layout.preferredLayoutMode,
      layout.setPreferredLayoutMode,
      layout.mode
    ]
  );
  const pageLayoutPortalContext = (0, import_react.useContext)(import_PageLayoutPortalContext.PageLayoutPortalContext);
  (0, import_useSafeLayoutEffect.useSafeLayoutEffect)(() => {
    if (!pageLayoutPortalContext) {
      return;
    }
    pageLayoutPortalContext.registry.register(
      "detailsLayoutContext",
      contextValue
    );
  }, [pageLayoutPortalContext, contextValue]);
  const contentNode = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      ref: content.ref,
      ...content.domProps,
      className: (0, import_clsx.clsx)(styles.content, content.domProps.className),
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(contentTeleport.Target, {})
    }
  );
  const layoutNode = (() => {
    switch (layout.mode) {
      case "collapsed": {
        return contentNode;
      }
      case "split": {
        return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_SplitLayout.SplitLayout, { children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_SplitLayout.SplitLayout.Content,
            {
              ...content.domProps,
              ref: content.ref,
              minWidth: content.config.minWidth,
              className: (0, import_clsx.clsx)(styles.content, content.domProps.className),
              children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(contentTeleport.Target, {})
            }
          ),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_SplitLayout.SplitLayout.Panel,
            {
              ...details.domProps,
              ref: details.ref,
              minWidth: details.config.minWidth,
              maxWidth: details.config.maxWidth,
              defaultWidth: persistedPanelWidthRef.current ?? details.config.defaultWidth,
              onResize: (value) => {
                persistedPanelWidthRef.current = value;
                details.config.onResize?.(value);
              },
              resizable: details.config.resizable,
              position: "end",
              className: (0, import_clsx.clsx)(styles.panel, details.domProps.className),
              "aria-label": detailsAriaLabel,
              children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(panelTeleport.Target, {})
            }
          )
        ] });
      }
      case "overlay": {
        if (layout.containerWidth === 0) {
          return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
            contentNode,
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { hidden: true, "aria-hidden": "true", children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(panelTeleport.Target, {}) })
          ] });
        }
        return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
          contentNode,
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_Overlay.Overlay,
            {
              ...details.domProps,
              ref: details.ref,
              minWidth: (0, import_width_utils.convertToPx)(
                details.config.minWidth,
                layout.containerWidth
              ),
              maxWidth: details.config.resizable ? layout.containerWidth - import_Overlay.OVERLAY_HANDLE_WIDTH : layout.containerWidth,
              defaultWidth: persistedPanelWidthRef.current ?? (0, import_width_utils.convertToPx)(details.config.defaultWidth, layout.containerWidth),
              resizable: details.config.resizable,
              onResize: (value) => {
                persistedPanelWidthRef.current = value;
                details.config.onResize?.(value);
              },
              children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(panelTeleport.Target, {})
            }
          )
        ] });
      }
      case "drawer": {
        return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
          contentNode,
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_Drawer.Drawer,
            {
              ...details.domProps,
              ref: details.ref,
              className: (0, import_clsx.clsx)(styles.drawer, details.domProps.className),
              "aria-label": detailsAriaLabel,
              isDismissed: layout.collapsed,
              onDismiss: (state) => layout.setCollapsed(state),
              placement: "right",
              width: "wide",
              children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: styles.drawerContent, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(panelTeleport.Target, {}) })
            }
          )
        ] });
      }
    }
  })();
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_DetailsLayoutContext.DetailsLayoutContext.Provider, { value: contextValue, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(panelTeleport.Source, { children: details.children }),
    layoutNode
  ] });
}
const DetailsLayout = (0, import_react.memo)(
  (0, import_react.forwardRef)((props, ref) => {
    const { className: rootClassName, details, content, ...rootProps } = props;
    const containerRef = (0, import_react.useRef)(null);
    const rootRef = (0, import_useMergeRefs.useMergeRefs)([ref, containerRef]);
    const persistedPanelWidth = (0, import_react.useRef)();
    const contentTeleport = (0, import_useTeleport.useTeleport)();
    const panelTeleport = (0, import_useTeleport.useTeleport)();
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(contentTeleport.Source, { children: content.children }),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        "div",
        {
          ...rootProps,
          ref: rootRef,
          className: (0, import_clsx.clsx)(styles.layoutRoot, rootClassName),
          children: details ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            DetailsLayoutInner,
            {
              details,
              content,
              containerRef,
              persistedPanelWidthRef: persistedPanelWidth,
              contentTeleport,
              panelTeleport
            }
          ) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "div",
            {
              ref: content.ref,
              ...content.domProps,
              className: (0, import_clsx.clsx)(styles.content, content.domProps.className),
              children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(contentTeleport.Target, {})
            }
          )
        }
      )
    ] });
  })
);
DetailsLayout.displayName = "DetailsLayout (forwardRef)";
