// @dynatrace/strato-components/layouts/page-layout/details-layout/DetailsControlBar.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DetailsControlBar_exports = {};
__export(DetailsControlBar_exports, {
  DetailsControlBar: () => DetailsControlBar
});
module.exports = __toCommonJS(DetailsControlBar_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = require("clsx");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var styles = __toESM(require("./DetailsControlBar.sty.js"));
var import_DetailsLayoutContext = require("./DetailsLayoutContext.js");
var import_Button = require("../../../buttons/button/Button.js");
var import_Divider = require("../../divider/Divider.js");
const DetailsControlBar = (props) => {
  const { children, className, ...restProps } = props;
  const context = (0, import_DetailsLayoutContext.useDetailsLayoutContext)();
  const intl = (0, import_react_intl.useIntl)();
  const hasCustomActions = import_react.Children.count(children) > 0;
  const closeLabel = intl.formatMessage({
    defaultMessage: "Close details",
    id: "UK0OoP7XF++evklW",
    description: "Label for the button that closes the details panel."
  });
  const switchToOverlayLabel = intl.formatMessage({
    defaultMessage: "Switch to overlay",
    id: "+mEjL2q8RACwzIMR",
    description: "Label for the button that switches the details panel to overlay mode."
  });
  const switchToSplitLabel = intl.formatMessage({
    defaultMessage: "Switch to side-by-side",
    id: "RST92XLweEz0Qott",
    description: "Label for the button that switches the details panel to split mode."
  });
  const handleClose = () => {
    context?.setCollapsed(true);
  };
  const handleLayoutToggle = () => {
    if (!context) {
      return;
    }
    context.setPreferredLayout(
      (prev) => prev === "split" ? "overlay" : "split"
    );
  };
  const showLayoutToggle = context?.mode ? ["overlay", "split"].includes(context.mode) : false;
  const isSplit = context?.preferredLayout === "split";
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { ...restProps, className: (0, import_clsx.clsx)(styles.controlBar, className), children: [
    hasCustomActions && /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: styles.customActions, children }),
    hasCustomActions && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Divider.Divider, { orientation: "vertical", as: "hr", flexItem: true }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { className: styles.defaultActions, children: [
      showLayoutToggle && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_Button.Button,
        {
          "aria-label": isSplit ? switchToOverlayLabel : switchToSplitLabel,
          onClick: handleLayoutToggle,
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: isSplit ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.OverlayRightIcon, {}) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.SplitPanelRightIcon, {}) })
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_Button.Button,
        {
          "aria-label": closeLabel,
          onClick: handleClose,
          "data-dt-name": "CloseX",
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.XmarkIcon, {}) })
        }
      )
    ] })
  ] });
};
DetailsControlBar.displayName = "PageLayout.Details.ControlBar";
