// @dynatrace/strato-components/layouts/page-layout/details-layout/Overlay.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Overlay_exports = {};
__export(Overlay_exports, {
  OVERLAY_HANDLE_WIDTH: () => OVERLAY_HANDLE_WIDTH,
  Overlay: () => Overlay
});
module.exports = __toCommonJS(Overlay_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("@use-gesture/react");
var import_clsx = require("clsx");
var import_react2 = require("react");
var styles = __toESM(require("./Overlay.sty.js"));
var import_useFocusRing = require("../../../core/styles/useFocusRing.js");
var import_width_utils = require("../../split-layout/width-utils.js");
const OVERLAY_HANDLE_WIDTH = 8;
const Overlay = (0, import_react2.memo)(
  (0, import_react2.forwardRef)((props, forwardedRef) => {
    const {
      children,
      className,
      style,
      minWidth = 200,
      maxWidth = Infinity,
      defaultWidth = 300,
      resizable = true,
      onResize,
      ...restProps
    } = props;
    const containerRef = (0, import_react2.useRef)(null);
    const handleRef = (0, import_react2.useRef)(null);
    const { focusProps, focusClassName } = (0, import_useFocusRing.useFocusRing)({
      focusWithin: true,
      ignoreModality: true
    });
    const dragStartWidth = (0, import_react2.useRef)(null);
    const clampContentWidth = (width) => (0, import_width_utils.clamp)(minWidth, width, maxWidth);
    const clampedDefaultWidth = clampContentWidth(defaultWidth);
    const [resizedWidth, setResizedWidth] = (0, import_react2.useState)(null);
    const contentWidth = resizedWidth === null ? clampedDefaultWidth : clampContentWidth(resizedWidth);
    const overlayWidth = resizable ? contentWidth + OVERLAY_HANDLE_WIDTH : contentWidth;
    (0, import_react.useGesture)(
      {
        onDragStart: () => {
          dragStartWidth.current = clampContentWidth(contentWidth);
        },
        onDragEnd: () => {
          dragStartWidth.current = null;
        },
        onDrag: (state) => {
          if (dragStartWidth.current === null) {
            return;
          }
          const nextWidth = (0, import_width_utils.clamp)(
            minWidth,
            dragStartWidth.current + state.movement[0] * -1,
            maxWidth
          );
          setResizedWidth(nextWidth);
          onResize?.(nextWidth);
        }
      },
      { target: handleRef }
    );
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      "div",
      {
        role: "dialog",
        className: styles.overlay,
        style: { ...style, width: overlayWidth },
        ref: containerRef,
        children: [
          resizable ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: styles.handleContainer, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "hr",
            {
              ...focusProps,
              ref: handleRef,
              className: (0, import_clsx.clsx)(styles.resizeHandle, focusClassName),
              tabIndex: 0
            }
          ) }) : null,
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "div",
            {
              ...restProps,
              ref: forwardedRef,
              className: (0, import_clsx.clsx)(styles.overlayContent, className),
              children
            }
          )
        ]
      }
    );
  })
);
Overlay.displayName = "Overlay";
