// @dynatrace/strato-components/tables/DataTable/features/RowDetails/details-row-model.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var details_row_model_exports = {};
__export(details_row_model_exports, {
  getRowDetailsRowModel: () => getRowDetailsRowModel
});
module.exports = __toCommonJS(details_row_model_exports);
var import_constants = require("../../constants.js");
var import_utils = require("../../hooks/useTable/utils.js");
var import_src = require("../../tanstack-table-core-override/src/index.js");
function getRowDetailsRowModel() {
  return (table) => {
    return (0, import_src.memo)(
      () => [table.getState().rowDetails, table.getPreRowDetailsRowModel()],
      (_rowDetails, rowModel) => {
        if (!rowModel.rows.length) {
          return rowModel;
        }
        const { rows, flatRows, rowsById } = rowModel;
        const detailsRowModel = expandRowDetails({
          rows,
          flatRows,
          rowsById
        });
        return detailsRowModel;
      },
      (0, import_utils.getMemoOptions)(table.options, "debugRows", "rowDetailsRowModel")
    );
  };
}
function expandRowDetails(rowModel) {
  const detailsRows = [];
  rowModel.rows.forEach((row, index) => {
    const parity = index % 2 === 0 ? "even" : "odd";
    row[import_constants.DATA_TABLE_DETAILS_ROW_PARITY] = parity;
    detailsRows.push(row);
    if (row.getIsDetailsOpen()) {
      const clonedRow = Object.assign(row.clone(), {
        id: `${row.id}${import_constants.DATA_TABLE_ROW_EXPANDED_ID_SUFFIX}`,
        [import_constants.DATA_TABLE_DETAILS_ROW_IDENTIFIER]: true,
        // Make sure that the cloned row will not be selectable.
        getCanSelect: () => false,
        // Make sure that the cloned row will not show up in the subrows
        getCanExpand: () => false,
        // Make sure that the cloned row will be activateable
        getCanActivateRow: () => false,
        // Delegate to the original row so the aria-controls id matches the button's aria-controls.
        getRowDetailsAriaControlsId: () => row.getRowDetailsAriaControlsId()
      });
      detailsRows.push(clonedRow);
      rowModel.rowsById[clonedRow.id] = clonedRow;
    }
  });
  return {
    rows: detailsRows,
    flatRows: rowModel.flatRows,
    rowsById: rowModel.rowsById
  };
}
