// @dynatrace/strato-components/tables/DataTable/features/RowDetails/RowDetails.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var RowDetails_exports = {};
__export(RowDetails_exports, {
  DataTableExpandableRowTemplate: () => DataTableExpandableRowTemplate,
  DataTableExpandableRowWrapper: () => DataTableExpandableRowWrapper,
  DataTableRowDetails: () => DataTableRowDetails,
  ROW_DETAILS_COLUMN: () => ROW_DETAILS_COLUMN,
  isRowDetailsColumnDef: () => isRowDetailsColumnDef,
  useRowDetails: () => useRowDetails
});
module.exports = __toCommonJS(RowDetails_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_react_is = require("react-is");
var import_borders = __toESM(require("@dynatrace/strato-design-tokens/borders"));
var import_spacings = __toESM(require("@dynatrace/strato-design-tokens/spacings"));
var import_strato_icons = require("@dynatrace/strato-icons");
var import_details_row_model = require("./details-row-model.js");
var import_RowDetails_css = require("./RowDetails.sty.js");
var import_Button = require("../../../../buttons/button/Button.js");
var import_useControllableState = require("../../../../core/hooks/useControllableState.js");
var import_DataTableDefaultCell = require("../../components/DataTableDefaultCell.js");
var import_constants = require("../../constants.js");
var import_DataTableContext = require("../../contexts/DataTableContext.js");
var import_utils = require("../../hooks/useTable/utils.js");
var import_messages = require("../../messages.js");
function DataTableExpandableRowTemplate(props) {
  return null;
}
function DataTableExpandableRowWrapper(props) {
  const {
    className: consumerClassName,
    "data-testid": dataTestId,
    style: consumerStyle,
    ...remainingProps
  } = props;
  const table = (0, import_DataTableContext.useDataTable)();
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      "data-testid": dataTestId,
      ...remainingProps,
      className: (0, import_clsx.default)(
        consumerClassName,
        (0, import_RowDetails_css.rowDetailsWrapperCSS)({
          rowDensity: table.getVisualVariantOptions().rowDensity
        })
      ),
      style: consumerStyle
    }
  );
}
function isDataTableExpandableRowTemplate(child) {
  return (0, import_react_is.isElement)(child) && child.type === DataTableExpandableRowTemplate;
}
const DataTableRowDetails = {
  getInitialState(state) {
    return {
      rowDetails: {},
      ...state
    };
  },
  createTable(table) {
    table.rowDetailsTableId = (0, import_lodash_es.uniqueId)("DataTable-");
    table.getPreRowDetailsRowModel = () => table.getPaginationRowModel();
    table.getRowDetailsRowModel = () => {
      if (!table._getRowDetailsRowModel && table.options.getRowDetailsRowModel) {
        table._getRowDetailsRowModel = table.options.getRowDetailsRowModel(table);
      }
      if (!table._getRowDetailsRowModel) {
        return table.getPreRowDetailsRowModel();
      }
      return table._getRowDetailsRowModel();
    };
    (0, import_utils.assignRowMethods)(table, {
      /**
       * Determines if the details row for a particular row is open.
       */
      getIsDetailsOpen() {
        const detailOpenState = table.getState().rowDetails ?? false;
        if (typeof detailOpenState === "boolean") {
          return detailOpenState;
        }
        return detailOpenState[this.id] === true;
      },
      /**
       * Defines the grid position for the row details row.
       */
      getRowDetailsGridPosition() {
        return {
          gridColumnStart: 1,
          gridColumnEnd: `span ${table.getVisibleLeafColumns().length}`
        };
      },
      /**
       * Returns the aria control id to connect the expanding button to the
       * row details wrapper.
       */
      getRowDetailsAriaControlsId() {
        return `${table.rowDetailsTableId}-${this.id}-expand`;
      },
      /**
       * Determines if the details for this row can be shown.
       */
      getCanOpenDetails() {
        if (typeof table.options.rowDetailsDisableRow === "function") {
          return !table.options.rowDetailsDisableRow(this.original);
        }
        return true;
      },
      /**
       * Control command to toggle the state of the rowDetails.
       */
      toggleDetails() {
        const scrollOffset = table.getRowScrollOffset();
        table.options.onRowDetailsOpenChange?.(
          (prev) => {
            let newRowDetailsState = {};
            if (typeof prev === "boolean" && prev === true) {
              const flatRows = table.getRowModel().flatRows;
              for (const flatRow of flatRows) {
                if (this.getCanOpenDetails() && this.id !== flatRow.id) {
                  newRowDetailsState[flatRow.id] = true;
                }
              }
              return newRowDetailsState;
            }
            newRowDetailsState = { ...prev };
            if (newRowDetailsState[this.id] === true) {
              delete newRowDetailsState[this.id];
            } else {
              newRowDetailsState[this.id] = true;
            }
            return newRowDetailsState;
          }
        );
        requestAnimationFrame(() => {
          table.scrollToRowOffset(scrollOffset);
        });
      }
    });
  }
};
function isRowDetailsColumnDef(column) {
  return column.id === ROW_DETAILS_COLUMN.id;
}
function RowDetailsCell(props) {
  const { row } = props;
  const intl = (0, import_react_intl.useIntl)();
  const isOpen = row.getIsDetailsOpen();
  const Icon = isOpen ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ChevronDownSmallIcon, {}) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ChevronRightSmallIcon, {});
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_DataTableDefaultCell.DataTableDefaultCell, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Button.Button,
    {
      size: "condensed",
      variant: "default",
      style: {
        borderRadius: import_borders.default.Radius.Field.Subdued,
        paddingLeft: import_spacings.default.Size2,
        paddingRight: import_spacings.default.Size2,
        paddingTop: 0,
        paddingBottom: 0
      },
      "aria-label": intl.formatMessage(import_messages.dataTableMessages.rowDetailsTrigger, {
        isDetailsOpen: isOpen,
        id: row.id
      }),
      "aria-expanded": isOpen,
      "aria-controls": isOpen ? row.getRowDetailsAriaControlsId() : void 0,
      onClick: () => row.toggleDetails(),
      disabled: !row.getCanOpenDetails(),
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: Icon })
    }
  ) });
}
const ROW_DETAILS_COLUMN = {
  id: import_constants.DATA_TABLE_ROW_DETAILS_COLUMN_ID,
  alignment: "left",
  accessorFn: () => "",
  header: () => {
    return "";
  },
  width: "content",
  enableSorting: false,
  enableResizing: false,
  enableUserActions: false,
  enableHiding: false,
  meta: {
    builtin: true,
    builtinPlacement: "prefix",
    hasCustomCellRenderer: true,
    hasCustomHeaderRenderer: true,
    cellDeps: (cell) => {
      return [cell.row.getIsDetailsOpen()];
    }
  },
  cell: ({ row }) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(RowDetailsCell, { row })
};
function useRowDetails(props, options) {
  const { children } = props;
  const rowDetailsTemplateProps = (0, import_react.useMemo)(() => {
    for (const child of import_react.Children.toArray(children)) {
      if (isDataTableExpandableRowTemplate(child)) {
        return child?.props;
      }
    }
  }, [children]);
  const [openRowDetails, setOpenRowDetails] = (0, import_useControllableState.useControllableState)({
    value: rowDetailsTemplateProps?.expandedRows,
    defaultValue: rowDetailsTemplateProps?.defaultExpandedRows,
    onChange: (value) => {
      if (value !== true) {
        rowDetailsTemplateProps?.onExpandedRowsChange?.(value);
      }
    }
  });
  const template = rowDetailsTemplateProps?.children;
  if (template !== void 0 && typeof template === "function") {
    options.enableRowDetails = true;
    options.rowDetailsTemplate = template;
    options.getRowDetailsRowModel = (0, import_details_row_model.getRowDetailsRowModel)();
    options.rowDetailsDisableRow = rowDetailsTemplateProps?.disableExpand;
    options.state = {
      ...options.state,
      rowDetails: openRowDetails
    };
    options.onRowDetailsOpenChange = setOpenRowDetails;
  }
}
