// @dynatrace/strato-components/tables/DataTable/features/RowDetails/RowDetailsElement.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var RowDetailsElement_exports = {};
__export(RowDetailsElement_exports, {
  RowDetailsElement: () => RowDetailsElement
});
module.exports = __toCommonJS(RowDetailsElement_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_RowDetails_css = require("./RowDetails.sty.js");
var import_data_table_border_width = require("../../../utils/data-table-border-width.js");
var import_DataTableContext = require("../../contexts/DataTableContext.js");
function RowDetailsElement(props) {
  const {
    originalRow,
    virtualRow,
    virtualItemIndex,
    rowSeparation,
    rowParity,
    rowMeasure,
    isLastRowOnPage: isLastRow,
    isDraggedRow = false,
    children
  } = props;
  const table = (0, import_DataTableContext.useDataTable)();
  let width = table.getRowVirtualizationContainer()?.offsetWidth;
  if (width && table.getVisualVariantOptions().contained) {
    width -= 2 * (0, import_data_table_border_width.getDataTableBorderWidth)();
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      role: "row",
      id: originalRow.getRowDetailsAriaControlsId(),
      style: {
        ...originalRow.getGridRowPosition(virtualItemIndex),
        ...originalRow.getRowDetailsGridPosition(),
        maxWidth: width,
        position: "sticky",
        left: 0
      },
      className: (0, import_RowDetails_css.rowDetailsRow)({
        rowSeparation,
        rowParity,
        isDraggedRow,
        isLastRow
      }),
      ref: rowMeasure,
      "data-rowid": originalRow.id,
      "data-row": rowMeasure ? virtualRow.index : void 0,
      "data-column": 0,
      children
    }
  );
}
