// @dynatrace/strato-components/charts/categorical-bar/CategoricalBarChart.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CategoricalBarChart_exports = {};
__export(CategoricalBarChart_exports, {
  CategoricalBarChart: () => CategoricalBarChart
});
module.exports = __toCommonJS(CategoricalBarChart_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_CategoricalBarChartCategoryAxis = require("./components/axes/CategoricalBarChartCategoryAxis.js");
var import_CategoricalBarChartValueAxis = require("./components/axes/CategoricalBarChartValueAxis.js");
var import_CategoricalBarChartOverlays = require("./components/interactions/CategoricalBarChartOverlays.js");
var import_CategoricalThresholdArea = require("./components/thresholds/CategoricalThresholdArea.js");
var import_CategoricalThresholdIndicators = require("./components/thresholds/CategoricalThresholdIndicators.js");
var import_constants = require("./constants.js");
var import_useCategoricalBarChartBaseConfig = require("./hooks/useCategoricalBarChartBaseConfig.js");
var import_useCategoricalBarChartCategoryAxisConfig = require("./hooks/useCategoricalBarChartCategoryAxisConfig.js");
var import_useCategoricalBarChartData = require("./hooks/useCategoricalBarChartData.js");
var import_useCategoricalBarChartLegend = require("./hooks/useCategoricalBarChartLegend.js");
var import_CategoricalBarChartPlot = require("./layout/CategoricalBarChartPlot.js");
var import_CategoricalBarChartFormatterProvider = require("./providers/CategoricalBarChartFormatterProvider.js");
var import_CategoricalBarScalesProviders = require("./providers/CategoricalBarScalesProviders.js");
var import_categorical_bar_area_rules = require("./types/categorical-bar-area-rules.js");
var import_layout = require("./utils/layout.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_defaultChartValues_css = require("../../core/styles/defaultChartValues.sty.js");
var import_ChartGridRoot = require("../core/components/chart-grid/ChartGridRoot.js");
var import_useChartLayoutLoading = require("../core/components/chart-layout/hooks/useChartLayoutLoading.js");
var import_useChartLayoutOrientation = require("../core/components/chart-layout/hooks/useChartLayoutOrientation.js");
var import_useGraphSize = require("../core/components/chart-layout/hooks/useGraphSize.js");
var import_constants2 = require("../core/components/data-point/constants.js");
const categoricalGridTemplateAreas = `
  "${import_categorical_bar_area_rules.AREAS.leftAxis} ${import_categorical_bar_area_rules.AREAS.leftThreshold} plot"
  ". . ${import_categorical_bar_area_rules.AREAS.bottomThreshold}"
  ". . ${import_categorical_bar_area_rules.AREAS.bottomAxis}"
`;
const categoricalGridTemplateColumns = (leftAxisMaxWidth = "auto") => {
  const max = typeof leftAxisMaxWidth === "number" ? `${leftAxisMaxWidth}px` : leftAxisMaxWidth;
  return `minmax(0, ${max}) auto minmax(60%, 1fr)`;
};
const categoricalGridTemplateRows = (bottomAxisMaxHeight = "auto") => {
  const max = typeof bottomAxisMaxHeight === "number" ? `${bottomAxisMaxHeight}px` : bottomAxisMaxHeight;
  return `minmax(60%, 1fr) auto minmax(0, ${max})`;
};
const CategoricalBarChart = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const loading = (0, import_useChartLayoutLoading.useChartLayoutLoading)();
  const chartRef = (0, import_react.useRef)(null);
  const chartAndForwardedRef = (0, import_useMergeRefs.useMergeRefs)([forwardedRef, chartRef]);
  const { maxSize } = (0, import_useCategoricalBarChartCategoryAxisConfig.useCategoricalBarChartCategoryAxisConfig)();
  const { layout } = (0, import_useCategoricalBarChartBaseConfig.useCategoricalBarChartBaseConfig)();
  const { hidden: legendHidden } = (0, import_useCategoricalBarChartLegend.useCategoricalBarChartLegend)();
  const { orientation } = (0, import_useChartLayoutOrientation.useChartLayoutOrientation)();
  const { uniqueDimensions } = (0, import_useCategoricalBarChartData.useCategoricalBarChartData)();
  const maxSizeLeftAxis = (0, import_layout.isVertical)(layout) ? void 0 : maxSize;
  const maxSizeBottomAxis = (0, import_layout.isVertical)(layout) ? maxSize : void 0;
  const { width, height } = (0, import_useGraphSize.useGraphSize)();
  const dontShowLegend = legendHidden || uniqueDimensions.length <= 1;
  const noLegendRight = dontShowLegend || orientation === "vertical";
  const noLegendBottom = dontShowLegend || orientation === "horizontal";
  const chartGridHeight = height - (noLegendBottom ? import_constants2.OVERLAY_PADDING + import_constants2.BOTTOM_PADDING_ADJUSTMENT : 0);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      style: { paddingTop: import_constants2.OVERLAY_PADDING },
      className: import_defaultChartValues_css.defaultChartSize,
      "aria-label": import_constants.CATEGORICAL_BAR_CHART_ARIA_LABEL,
      role: "img",
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_ChartGridRoot.ChartGrid,
        {
          width: width - (noLegendRight ? import_constants2.OVERLAY_PADDING : 0),
          height: chartGridHeight,
          gridTemplateAreas: categoricalGridTemplateAreas,
          gridTemplateColumns: categoricalGridTemplateColumns(maxSizeLeftAxis),
          gridTemplateRows: categoricalGridTemplateRows(maxSizeBottomAxis),
          ref: chartAndForwardedRef,
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_CategoricalBarScalesProviders.CategoricalBarScalesProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_CategoricalBarChartFormatterProvider.CategoricalBarChartFormatterProvider, { children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_CategoricalBarChartCategoryAxis.CategoricalBarChartCategoryAxis, {}),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_CategoricalBarChartValueAxis.CategoricalBarChartValueAxis, {}),
            /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_ChartGridRoot.ChartGrid.PlotArea, { children: [
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_CategoricalBarChartPlot.CategoricalBarChartPlot, {}),
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_CategoricalThresholdIndicators.CategoricalThresholdIndicators, {}),
              loading ? null : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_CategoricalBarChartOverlays.CategoricalBarChartOverlays, {})
            ] }),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_CategoricalThresholdArea.CategoricalThresholdArea, {})
          ] }) })
        }
      )
    }
  );
});
CategoricalBarChart.displayName = import_constants.DISPLAY_NAME;
