// @dynatrace/strato-components/charts/categorical-bar/CategoricalBarChartRoot.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CategoricalBarChartRoot_exports = {};
__export(CategoricalBarChartRoot_exports, {
  CategoricalBarChart: () => CategoricalBarChart,
  CategoricalBarChartRoot: () => CategoricalBarChartRoot
});
module.exports = __toCommonJS(CategoricalBarChartRoot_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_CategoricalBarChartLayout = require("./CategoricalBarChartLayout.js");
var import_constants = require("./constants.js");
var import_CategoricalBarChartBaseConfig_context = require("./contexts/CategoricalBarChartBaseConfig.context.js");
var import_CategoricalBarChartRawData_context = require("./contexts/CategoricalBarChartRawData.context.js");
var import_selectors = require("./hooks/store/selectors.js");
var import_useCategoricalBarChartConfig = require("./hooks/useCategoricalBarChartConfig.js");
var import_CategoricalBarChartProviders = require("./providers/CategoricalBarChartProviders.js");
var import_CategoricalBarDataProvider = require("./providers/CategoricalBarDataProvider.js");
var import_CategoricalBarFilteredDataProvider = require("./providers/CategoricalBarFilteredDataProvider.js");
var import_StoreProvider = require("./providers/StoreProvider.js");
var import_CategoryAxis = require("./slots/CategoryAxis.js");
var import_DownloadCSV = require("./slots/DownloadCSV.js");
var import_Legend = require("./slots/Legend.js");
var import_ThresholdIndicator = require("./slots/ThresholdIndicator.js");
var import_Toolbar = require("./slots/Toolbar.js");
var import_Tooltip = require("./slots/Tooltip.js");
var import_ValueAxis = require("./slots/ValueAxis.js");
var import_defaults = require("./types/defaults.js");
var import_state = require("./types/state.js");
var import_build_and_download_csv = require("./utils/build-and-download-csv.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_Intent = require("../../core/slots/Intent/Intent.js");
var import_useAutoLegendRefresh = require("../core/components/chart-layout/hooks/useAutoLegendRefresh.js");
var import_get_chart_states_template = require("../core/components/chart-states/utils/get-chart-states-template.js");
var import_OverlayContext = require("../core/components/overlay-tooltip/contexts/OverlayContext.js");
var import_TooltipContext = require("../core/components/overlay-tooltip/contexts/TooltipContext.js");
var import_ChartSeriesActionContext = require("../core/components/series-action/ChartSeriesActionContext.js");
var import_ColorRuleContext = require("../core/context/ColorRuleContext.js");
var import_EmptyStateSlot = require("../core/slots/chart-states/EmptyStateSlot.js");
var import_ErrorStateSlot = require("../core/slots/chart-states/ErrorStateSlot.js");
var import_ColorRule = require("../core/slots/ColorRule.js");
var import_chart_types = require("../core/types/chart-types.js");
var import_coerceSizeValue = require("../core/utils/coerceSizeValue.js");
const COMPONENT_NAME_BEHAVIOURAL_TRACKING = "CategoricalBarChart";
const CategoricalBarChartRoot = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    data,
    height: heightProp,
    width: widthProp,
    seriesActions,
    loading = false,
    className,
    style,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow
  } = props;
  const [behaviorTrackingProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(props);
  const chartHeight = (0, import_coerceSizeValue.coerceSizeValue)(heightProp ?? import_defaults.DEFAULT_CHART_HEIGHT);
  const chartConfig = (0, import_useCategoricalBarChartConfig.useCategoricalBarChartConfig)(props);
  const { emptyState, errorState } = (0, import_get_chart_states_template.getChartStatesTemplates)(props.children);
  const containerRef = (0, import_react.useRef)(null);
  const rootRef = (0, import_useMergeRefs.useMergeRefs)([
    forwardedRef,
    containerRef
  ]);
  (0, import_react.useImperativeHandle)(forwardedRef, () => ({
    element: containerRef.current,
    getConfig: () => JSON.stringify({
      type: import_chart_types.ChartType.CategoricalBar,
      ...chartConfig
    }),
    downloadData: () => {
      (0, import_build_and_download_csv.buildAndDownloadCsv)(data);
    }
  }));
  const baseConfig = {
    valueRepresentation: chartConfig.valueRepresentation,
    colorPalette: chartConfig.colorPalette,
    layout: chartConfig.layout,
    groupMode: chartConfig.groupMode,
    colorPaletteMode: chartConfig.colorPaletteMode,
    toolbar: chartConfig.toolbar,
    intents: chartConfig.intents
  };
  const setState = (0, import_selectors.useSetState)();
  (0, import_react.useEffect)(() => {
    setState(() => import_state.initialState);
  }, [chartConfig]);
  const chartLayoutRef = (0, import_useAutoLegendRefresh.useAutoLegendRefresh)(data);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_CategoricalBarChartBaseConfig_context.CategoricalBarChartBaseConfigContext.Provider, { value: baseConfig, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ColorRuleContext.ColorRuleContext.Provider, { value: chartConfig.colorRules, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_CategoricalBarChartRawData_context.CategoricalBarChartRawDataContext.Provider, { value: data, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_CategoricalBarDataProvider.CategoricalBarDataProvider, { data, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_StoreProvider.StoreProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ChartSeriesActionContext.ChartSeriesActionContext.Provider,
    {
      value: {
        // We need a cast here as the provider expects unknown and the
        // react context will not be able to carry the generic anyways.
        chartSeriesActions: seriesActions
      },
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_CategoricalBarChartProviders.CategoricalBarChartProviders, { config: chartConfig, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_CategoricalBarFilteredDataProvider.CategoricalBarFilteredDataProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayContext.OverlayProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TooltipContext.TooltipProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_CategoricalBarChartLayout.CategoricalBarChartLayout,
        {
          chartLayoutRef,
          ref: rootRef,
          height: chartHeight,
          width: widthProp,
          emptyState,
          errorState,
          isEmpty: (0, import_lodash_es.isEmpty)(data),
          loading,
          legend: chartConfig.legend,
          className,
          style,
          ...behaviorTrackingProps,
          "data-dtrum-mask": dataDtrumMask,
          "data-dtrum-allow": dataDtrumAllow,
          "data-dt-component": COMPONENT_NAME_BEHAVIOURAL_TRACKING
        }
      ) }) }) }) })
    }
  ) }) }) }) }) });
});
CategoricalBarChartRoot.displayName = import_constants.DISPLAY_NAME;
const CategoricalBarChart = Object.assign(CategoricalBarChartRoot, {
  Legend: import_Legend.Legend,
  Tooltip: import_Tooltip.Tooltip,
  CategoryAxis: import_CategoryAxis.CategoryAxis,
  ValueAxis: import_ValueAxis.ValueAxis,
  ThresholdIndicator: import_ThresholdIndicator.ThresholdIndicator,
  Intent: import_Intent.Intent,
  ColorRule: import_ColorRule.ColorRule,
  EmptyState: import_EmptyStateSlot.EmptyStateSlot,
  ErrorState: import_ErrorStateSlot.ErrorStateSlot,
  Toolbar: import_Toolbar.CategoricalBarChartToolbarSlot,
  DownloadCSV: import_DownloadCSV.CategoricalBarChartDownloadCSVSlot
});
