// @dynatrace/strato-components/charts/categorical-bar/CategoricalBarChartLayout.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CategoricalBarChartLayout_exports = {};
__export(CategoricalBarChartLayout_exports, {
  CategoricalBarChartLayout: () => CategoricalBarChartLayout
});
module.exports = __toCommonJS(CategoricalBarChartLayout_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_CategoricalBarChart = require("./CategoricalBarChart.js");
var import_CategoricalBarChartLegend = require("./components/legend/CategoricalBarChartLegend.js");
var import_useCategoricalBarChartBaseConfig = require("./hooks/useCategoricalBarChartBaseConfig.js");
var import_useCategoricalBarChartData = require("./hooks/useCategoricalBarChartData.js");
var import_layout = require("./utils/layout.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_ChartLayout = require("../core/components/chart-layout/ChartLayout.js");
const CategoricalBarChartLayout = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    height,
    width,
    legend,
    isEmpty,
    loading,
    errorState,
    emptyState,
    className,
    style,
    chartLayoutRef,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "data-dt-component": dataDtComponent
  } = props;
  const [behaviorTrackingProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(props);
  const { uniqueDimensions } = (0, import_useCategoricalBarChartData.useCategoricalBarChartData)();
  const { groupMode } = (0, import_useCategoricalBarChartBaseConfig.useCategoricalBarChartBaseConfig)();
  const canShowLegend = (0, import_layout.isGrouped)(groupMode) || (0, import_layout.isStacked)(groupMode) && uniqueDimensions.length > 1;
  const {
    position: legendPosition,
    // eslint-disable-next-line @typescript-eslint/no-deprecated
    ratio,
    initialRatio,
    onRatioChange
  } = legend;
  const legendRatio = initialRatio ?? ratio;
  const showLegend = canShowLegend && !legend.hidden;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_ChartLayout.ChartLayout,
    {
      ref: chartLayoutRef,
      chartHeight: height,
      chartWidth: width,
      isEmpty,
      emptyState,
      errorState,
      showLoader: loading,
      className,
      style,
      ...behaviorTrackingProps,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "data-dt-component": dataDtComponent,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ChartLayout.ChartLayout.Graph, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_CategoricalBarChart.CategoricalBarChart, { ref: forwardedRef }) }),
        showLegend && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_ChartLayout.ChartLayout.Legend,
          {
            position: legendPosition,
            initialRatio: legendRatio,
            onResize: onRatioChange,
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_CategoricalBarChartLegend.CategoricalBarChartLegend, {})
          }
        )
      ]
    }
  );
});
CategoricalBarChartLayout.displayName = "CategoricalBarChartLayout";
