// @dynatrace/strato-components/charts/categorical-bar/constants.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var constants_exports = {};
__export(constants_exports, {
  CATEGORICAL_BAR_CHART_ARIA_LABEL: () => CATEGORICAL_BAR_CHART_ARIA_LABEL,
  DEFAULT_SCALE_FALLBACK_COLOR: () => DEFAULT_SCALE_FALLBACK_COLOR,
  DISPLAY_NAME: () => DISPLAY_NAME,
  VARIANT_GROUPED: () => VARIANT_GROUPED,
  zIndicesCategoricalAreas: () => zIndicesCategoricalAreas
});
module.exports = __toCommonJS(constants_exports);
var import_constants = require("../core/constants.js");
const zIndicesCategoricalAreas = {
  magnifier: 1,
  dataPoints: 2,
  toolbar: 3
};
const DEFAULT_SCALE_FALLBACK_COLOR = import_constants.SCALE_FALLBACK_COLOR;
const VARIANT_GROUPED = "grouped";
const CATEGORICAL_BAR_CHART_ARIA_LABEL = "Categorical Bar Chart";
const DISPLAY_NAME = "CategoricalBarChart";
