// @dynatrace/strato-components/esm/charts/core/components/tooltip/section-components/TooltipItem/TooltipItem.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { useMemo } from "react";
import Spacings from "@dynatrace/strato-design-tokens/spacings";
import { container, itemContainer } from "./TooltipItem.sty.js";
import { useId } from "../../../../../../core/hooks/useId.js";
import { Divider } from "../../../../../../layouts/divider/Divider.js";
import { ChartActionsMenu } from "../../../series-action/ChartActionsMenu.js";
import { useActionSelection } from "../../context/ActionSelectionContext.js";
const TooltipItem = (props) => {
  const { children, showDivider, actions, defaultActions, ...rest } = props;
  const itemId = useId();
  const { selectedItemId, actionsMenuRef, updateSelectedItem } = useActionSelection();
  const isOpen = selectedItemId === itemId;
  const memoizedActions = useMemo(() => {
    if (isOpen && actions) {
      return actions();
    }
  }, [isOpen]);
  if (!actions && !defaultActions) {
    return /* @__PURE__ */ jsxs(Fragment, { children: [
      /* @__PURE__ */ jsx("div", { className: itemContainer, ...rest, children }),
      showDivider && /* @__PURE__ */ jsx(
        Divider,
        {
          orientation: "horizontal",
          style: { margin: `${Spacings.Size4} 0` }
        }
      )
    ] });
  }
  return /* @__PURE__ */ jsx(
    ChartActionsMenu,
    {
      isOpen,
      actions: memoizedActions,
      defaultActions,
      onActionClick: () => updateSelectedItem(""),
      ref: actionsMenuRef,
      children: /* @__PURE__ */ jsxs(
        "button",
        {
          className: container,
          onMouseEnter: () => updateSelectedItem(itemId),
          children: [
            /* @__PURE__ */ jsx("div", { className: itemContainer, ...rest, children }),
            showDivider && /* @__PURE__ */ jsx(
              Divider,
              {
                orientation: "horizontal",
                style: { margin: `${Spacings.Size4} 0` }
              }
            )
          ]
        }
      )
    }
  );
};
TooltipItem["displayName"] = "TooltipItem";
export {
  TooltipItem
};
