// @dynatrace/strato-geo/esm/map/components/DotLayer/DotLayer.js
import { jsx, jsxs } from "react/jsx-runtime";
import { Layer, Source } from "@vis.gl/react-maplibre";
import { _useBehavioralTrackingProps as useBehavioralTrackingProps } from "@dynatrace/strato-components/core";
import { BackgroundLayer } from "./BackgroundLayer.js";
import { BackgroundSymbolLayer } from "./BackgroundSymbolLayer.js";
import { DotLayerTooltip } from "./DotLayerTooltip.js";
import { parseDotDataToGeoJSON } from "./utils/parse-dot-data-to-geo-json.js";
import {
  DEFAULT_ICON_COLOR,
  DEFAULT_ICON_SIZE_RATIO,
  DEFAULT_SHAPE_COLOR,
  DEFAULT_SYMBOL_SIZE,
  FALLBACK_SYMBOL_SIZE,
  SHAPE_OPACITY_DIMMED,
  SHAPE_OPACITY_DEFAULT
} from "../../constants.js";
import { useLayerColoringStrategy } from "../../hooks/use-layer-coloring-strategy.js";
import { useLegendInteractionHighlighting } from "../../hooks/use-legend-interaction-highlighting.js";
import { useLegendInteractionVisibility } from "../../hooks/use-legend-interaction-visibility.js";
import { useLoadSymbolIntoMap } from "../../hooks/use-load-symbol-into-map.js";
import { useResolveLocationColor } from "../../hooks/use-resolve-color.js";
import { useTooltipTemplate } from "../../hooks/use-tooltip-template.js";
import { DotLayerTooltip as DotLayerTooltipSlot } from "../../slots/DotLayerTooltip.js";
import { isDotTooltipTemplate } from "../../utils/tooltip-type-guards.js";
import { TooltipWrapper } from "../TooltipWrapper.js";
const DotLayer = (props) => {
  const [behavioralTrackingProps] = useBehavioralTrackingProps(props);
  const {
    data,
    layerId,
    bearing = 0,
    children,
    shape = "pin",
    background = false,
    shapeSize = DEFAULT_SYMBOL_SIZE
  } = props;
  const calculatedShapeSize = shapeSize > 0 ? shapeSize : FALLBACK_SYMBOL_SIZE;
  const { symbolName, symbolType } = useLoadSymbolIntoMap(
    shape,
    layerId,
    calculatedShapeSize
  );
  const bgLayerId = `${layerId}-background`;
  const { symbolName: bgName } = useLoadSymbolIntoMap(
    "circle",
    bgLayerId,
    calculatedShapeSize
  );
  const defaultSymbolColor = symbolType === "shape" ? DEFAULT_SHAPE_COLOR : DEFAULT_ICON_COLOR;
  const colorParser = useLayerColoringStrategy();
  const shapeColor = useResolveLocationColor(
    defaultSymbolColor,
    props,
    colorParser
  );
  const {
    handler: tooltipTemplate,
    hidden: tooltipHidden,
    symbolAlignment
  } = useTooltipTemplate(children, DotLayerTooltipSlot, { extractProps: true });
  const sourceId = `source-${layerId}`;
  useLegendInteractionHighlighting(props, sourceId);
  useLegendInteractionVisibility(props, layerId);
  const dataGeoJson = parseDotDataToGeoJSON(
    data,
    shapeColor,
    bearing,
    background
  );
  const iconAnchor = shape === "pin" ? "bottom" : "center";
  return /* @__PURE__ */ jsxs(Source, { id: sourceId, type: "geojson", data: dataGeoJson, generateId: true, children: [
    /* @__PURE__ */ jsx(
      Layer,
      {
        type: "symbol",
        id: layerId,
        layout: {
          "icon-image": symbolName,
          "icon-size": DEFAULT_ICON_SIZE_RATIO,
          "icon-rotate": ["get", "__bearing"],
          // https://maplibre.org/maplibre-style-spec/layers/#layout-symbol-icon-ignore-placement
          "icon-ignore-placement": true,
          "icon-anchor": iconAnchor
        },
        paint: {
          "icon-color": ["get", "__color"],
          "icon-opacity": [
            "case",
            ["boolean", ["feature-state", "legendHover"], true],
            SHAPE_OPACITY_DEFAULT,
            SHAPE_OPACITY_DIMMED
          ]
        },
        ...behavioralTrackingProps
      }
    ),
    shape !== "pin" ? /* @__PURE__ */ jsx(
      BackgroundLayer,
      {
        background: Boolean(background),
        beforeId: layerId,
        source: sourceId,
        radius: calculatedShapeSize / 2,
        iconAnchor
      }
    ) : /* @__PURE__ */ jsx(
      BackgroundSymbolLayer,
      {
        background: Boolean(background),
        beforeId: layerId,
        source: sourceId,
        radius: calculatedShapeSize / 2,
        iconAnchor,
        symbolName: bgName,
        layerId: bgLayerId,
        shape
      }
    ),
    /* @__PURE__ */ jsx(TooltipWrapper, { layerId, children: /* @__PURE__ */ jsx(
      DotLayerTooltip,
      {
        tooltipTemplate: isDotTooltipTemplate(tooltipTemplate) ? tooltipTemplate : void 0,
        hidden: tooltipHidden,
        symbolAlignment
      }
    ) })
  ] });
};
DotLayer["displayName"] = "DotLayer";
export {
  DotLayer
};
