// @dynatrace/strato-components/forms/number-input-v2/NumberInput.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var NumberInput_exports = {};
__export(NumberInput_exports, {
  NumberInput: () => NumberInput
});
module.exports = __toCommonJS(NumberInput_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_calculations = require("./utils/calculations.js");
var import_format = require("./utils/format.js");
var import_validation = require("./utils/validation.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useId = require("../../core/hooks/useId.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_merge_props = require("../../core/utils/merge-props.js");
var import_BaseInput = require("../_base-input/BaseInput.js");
var import_BaseInputButton = require("../_base-input/BaseInputButton.js");
var import_BaseInputContent = require("../_base-input/BaseInputContent.js");
var import_useFormFieldMessagesContext = require("../forms-core/hooks/useFormFieldMessagesContext.js");
var import_formfield_messages_state_reducer = require("../forms-core/state/formfield-messages-state-reducer.js");
var import_helpers = require("../forms-core/time/helpers.js");
const COMPONENT_NAME = "NumberInput";
const VALID_PATTERN = "^[+\\-]?\\d+(?:[\\.,]\\d+)?(?:[eE][+\\-]?\\d+)?$";
const VALID_REGEXP = new RegExp(VALID_PATTERN);
const FORMATTABLE_PATTERN = /^[-+]?[\d.,'\u2019 ٫٬\u00A0\u1680\u2000-\u200A\u202F\u205F\u3000]+(?:[eE][-+]?\d+)?$/u;
const INTERMEDIATE_VALID_PATTERN = /^[+-]?\d*(?:[.,]\d*)?(?:[eE][+-]?\d*)?$/;
const NumberInputRoot = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const intl = (0, import_react_intl.useIntl)();
  const defaultPlaceholder = intl.formatMessage({
    defaultMessage: "Add number",
    id: "rrkUQoVNb46Jbd/8",
    description: "Default placeholder for the number input."
  });
  const {
    onChange,
    onKeyDown,
    value: propValue,
    defaultValue,
    required,
    readOnly = false,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "aria-disabled": ariaDisabledProp,
    min,
    max,
    step,
    stepMultiplier = 1,
    placeholder = defaultPlaceholder,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const { state, dispatch } = (0, import_useFormFieldMessagesContext.useFormFieldMessagesContext)();
  const errorId = (0, import_useId.useId)();
  const isPasted = (0, import_react.useRef)(false);
  const isInternalChanged = (0, import_react.useRef)(defaultValue !== void 0);
  const internalRef = (0, import_react.useRef)(null);
  const inputRef = (0, import_useMergeRefs.useMergeRefs)([forwardedRef, internalRef]);
  const [internalValue, setInternalValue] = (0, import_react.useState)(
    defaultValue ?? propValue ?? null
  );
  const setValueNative = (0, import_react.useCallback)(
    (newValue, triggerChange = true) => {
      const input = internalRef.current?.inputRef;
      if (input) {
        let stringValue;
        if (newValue === null) {
          stringValue = "";
        } else if (typeof newValue === "number") {
          stringValue = new Intl.NumberFormat((0, import_helpers.getSafeRegionalFormat)(), {
            useGrouping: false,
            maximumFractionDigits: 17
          }).format(newValue);
        } else {
          stringValue = String(newValue);
        }
        const desc = Object.getOwnPropertyDescriptor(
          HTMLInputElement.prototype,
          "value"
        );
        desc?.set?.call(input, stringValue);
        input.value = stringValue;
        if (triggerChange) {
          const parsedValue = (0, import_format.checkFormat)(stringValue);
          isInternalChanged.current = true;
          setInternalValue(parsedValue || stringValue);
          if (VALID_REGEXP.exec(parsedValue)) {
            onChange?.((0, import_format.parseNumber)(parsedValue));
          } else if (parsedValue === "") {
            onChange?.(null);
          }
        }
      }
    },
    [onChange]
  );
  const value = (0, import_react.useMemo)(() => {
    const isChanged = isInternalChanged.current;
    isInternalChanged.current = false;
    if (isChanged) {
      return internalValue;
    }
    if (!propValue && propValue !== 0) {
      return "";
    }
    const formattedPropValue = new Intl.NumberFormat((0, import_helpers.getSafeRegionalFormat)(), {
      useGrouping: false,
      maximumFractionDigits: 17
      // Max precision for a JS number
    }).format(propValue);
    return formattedPropValue;
  }, [propValue, internalValue]);
  const handleChange = (0, import_react.useCallback)(
    (e) => {
      const parsedValue = (0, import_format.checkFormat)(e.target.value);
      if (VALID_REGEXP.exec(parsedValue)) {
        onChange?.((0, import_format.parseNumber)(parsedValue));
      }
      if (parsedValue === "") {
        onChange?.(null);
      }
      if (isPasted.current && FORMATTABLE_PATTERN.exec(parsedValue)) {
        isInternalChanged.current = true;
        setInternalValue(parsedValue);
        setValueNative(parsedValue, false);
      } else if (INTERMEDIATE_VALID_PATTERN.exec(e.target.value)) {
        isInternalChanged.current = true;
        setInternalValue(parsedValue);
      } else {
        setValueNative(value ?? "", false);
      }
      isPasted.current = false;
    },
    [onChange, setInternalValue, value, setValueNative]
  );
  const handleStep = (0, import_react.useCallback)(
    (direction) => {
      setValueNative(
        (0, import_calculations.applyStep)(value, step, stepMultiplier, min, max, direction)
      );
    },
    [value, stepMultiplier, step, min, max, setValueNative]
  );
  const handleKeyDown = (0, import_react.useCallback)(
    (e) => {
      if (readOnly) {
        onKeyDown?.(e);
        return;
      }
      if (e.key === "ArrowUp") {
        e.preventDefault();
        handleStep(1);
      }
      if (e.key === "ArrowDown") {
        e.preventDefault();
        handleStep(-1);
      }
      if (e.key === "Home" && min !== void 0) {
        e.preventDefault();
        setValueNative(min);
      }
      if (e.key === "End" && max !== void 0) {
        e.preventDefault();
        setValueNative(max);
      }
      onKeyDown?.(e);
    },
    [handleStep, onKeyDown, setValueNative, min, max, readOnly]
  );
  (0, import_react.useEffect)(() => {
    const el = internalRef?.current;
    if (!el) {
      return;
    }
    const onWheel = (e) => {
      if (document.activeElement !== e.target) {
        return;
      }
      e.preventDefault();
      handleStep(e.deltaY < 0 ? 1 : -1);
    };
    el.element?.addEventListener("wheel", onWheel, { passive: false });
    return () => el.element?.removeEventListener("wheel", onWheel);
  }, [handleStep]);
  (0, import_react.useEffect)(() => {
    if (!value && value !== 0) {
      return;
    }
    const errorMessage = (0, import_validation.getErrorKey)(value, min, max, step);
    if (errorMessage) {
      (0, import_validation.addError)(errorMessage, errorId, internalRef.current?.inputRef, dispatch);
      return;
    }
    dispatch?.({
      type: import_formfield_messages_state_reducer.FormFieldMessagesAction.REMOVEMESSAGE,
      payload: {
        id: errorId,
        messageType: "formcontrol"
      }
    });
  }, [
    value,
    min,
    max,
    step,
    state.isTouched,
    state.isValid,
    dispatch,
    errorId
  ]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_BaseInput.BaseInput,
    {
      ref: inputRef,
      ...(0, import_merge_props.mergeProps)(restProps, behavioralTrackingProps),
      type: "text",
      inputMode: "decimal",
      autoComplete: "off",
      value: value ?? "",
      required,
      readOnly,
      onChange: handleChange,
      onKeyDown: handleKeyDown,
      onPaste: () => isPasted.current = true,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "aria-disabled": ariaDisabledProp,
      "data-dt-component": COMPONENT_NAME,
      min,
      max,
      step,
      pattern: VALID_PATTERN,
      placeholder
    }
  );
});
const NumberInput = Object.assign(NumberInputRoot, {
  Prefix: import_BaseInputContent.BaseInputPrefix,
  Suffix: import_BaseInputContent.BaseInputSuffix,
  Button: import_BaseInputButton.BaseInputButton
});
NumberInput.displayName = COMPONENT_NAME;
