// @dynatrace/strato-components/content/chip/BaseChip.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var BaseChip_exports = {};
__export(BaseChip_exports, {
  BaseChip: () => BaseChip
});
module.exports = __toCommonJS(BaseChip_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_is = require("react-is");
var styles = __toESM(require("./BaseChip.sty.js"));
var import_ChipContent = require("./ChipContent.js");
var import_ChipDeleteButton = require("./ChipDeleteButton.js");
var import_ChipKey = require("./ChipKey.js");
var import_ChipContext = require("./contexts/ChipContext.js");
var import_useChipContext = require("./contexts/useChipContext.js");
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useId = require("../../core/hooks/useId.js");
var import_useSafeForwardProps = require("../../core/hooks/useSafeForwardProps.js");
var import_focusRingStatic_css = require("../../core/styles/focusRingStatic.sty.js");
var import_Flex = require("../../layouts/flex/Flex.js");
var import_field_css = require("../../styles/field.sty.js");
var import_textStyle_css = require("../../styles/textStyle.sty.js");
var import_Text = require("../../typography/text/Text.js");
const COMPONENT_NAME = "Chip";
const getChipSlots = (node) => {
  const slots = {
    deleteButton: null,
    prefixIcons: [],
    suffixIcons: [],
    value: [],
    key: []
  };
  for (const child of import_react.Children.toArray(node)) {
    if ((0, import_react_is.isElement)(child)) {
      switch (child.type) {
        case import_ChipContent.ChipPrefix:
          slots.prefixIcons.push(child);
          break;
        case import_ChipContent.ChipSuffix:
          slots.suffixIcons.push(child);
          break;
        case import_ChipKey.ChipKey:
          slots.key.push(child);
          break;
        case import_ChipDeleteButton.ChipDeleteButton:
          slots.deleteButton = child;
          break;
        default:
          slots.value.push(child);
      }
    } else {
      slots.value.push(child);
    }
  }
  return slots;
};
const RootBaseChip = /* @__PURE__ */ (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const {
      children,
      as = "div",
      variant = "emphasized",
      color = "neutral",
      size = "default",
      disabled = false,
      maxWidth,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "data-dt-component": dataDtComponent = COMPONENT_NAME,
      ...remainingProps
    } = props;
    const ariaLabelingProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(remainingProps);
    const forwardProps = (0, import_useSafeForwardProps.useSafeForwardProps)(remainingProps, ariaLabelingProps);
    const [behavioralTrackingProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
    const { prefixIcons, suffixIcons, deleteButton, key, value } = (0, import_react.useMemo)(
      () => getChipSlots(children),
      [children]
    );
    const valueId = (0, import_useId.useId)();
    const hasKey = key.length > 0;
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Flex.Flex,
      {
        as: "div",
        ref: forwardedRef,
        gap: 2,
        className: (0, import_clsx.default)(
          styles.wrapper,
          disabled && styles.disabledWrapper,
          (0, import_field_css.fieldTokens)({ variant, color, disabled }),
          consumerClassName
        ),
        style: {
          maxWidth: maxWidth ?? 250,
          ...consumerStyle
        },
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        ...behavioralTrackingProps,
        "data-dt-component": dataDtComponent,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
          import_ChipContext.ChipContext.Provider,
          {
            value: { disabled, variant, color, size, valueId, hasKey },
            children: [
              /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
                import_Text.Text,
                {
                  as,
                  textStyle: "base",
                  className: (0, import_clsx.default)(
                    "strato-surface",
                    (0, import_focusRingStatic_css.focusRingStatic)({ color, disabled, when: "focus-visible" }),
                    (as === "a" || as === "button") && (0, import_textStyle_css.textStyle)({ textStyle: "base-emphasized" }),
                    (0, import_field_css.fieldTokens)({ variant, color, disabled }),
                    styles.container({
                      size,
                      disabled,
                      color,
                      variant,
                      hasDeleteButton: deleteButton !== null
                    })
                  ),
                  size,
                  disabled,
                  variant,
                  color,
                  ...ariaLabelingProps,
                  ...forwardProps,
                  children: [
                    prefixIcons.map((node, index) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react.Fragment, { children: node }, `prefix-node-${index}`)),
                    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(TextContainer, { keys: key, values: value }),
                    suffixIcons.map((node, index) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react.Fragment, { children: node }, `suffix-node-${index}`))
                  ]
                }
              ),
              deleteButton
            ]
          }
        )
      }
    );
  }
);
function TextContainer({
  keys,
  values
}) {
  const { valueId: id } = (0, import_useChipContext.useChipContext)();
  const hasValue = values.length > 0;
  const hasKey = keys.length > 0;
  if (!(hasKey || hasValue)) {
    return null;
  }
  const separator = hasKey && hasValue ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { children: ":" }) : null;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { className: styles.textContainer, children: [
    keys.map((node, index) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react.Fragment, { children: node }, `key-node-${index}`)),
    separator,
    values.map((node, index) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Flex.Flex, { id, className: styles.value, children: node }, `value-node-${index}`))
  ] });
}
const BaseChip = Object.assign(RootBaseChip, {
  Prefix: import_ChipContent.ChipPrefix,
  Suffix: import_ChipContent.ChipSuffix,
  Key: import_ChipKey.ChipKey,
  DeleteButton: import_ChipDeleteButton.ChipDeleteButton
});
BaseChip.displayName = COMPONENT_NAME;
