// @dynatrace/strato-components/content/chip/ChipContent.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ChipContent_exports = {};
__export(ChipContent_exports, {
  ChipContent: () => ChipContent,
  ChipPrefix: () => ChipPrefix,
  ChipSuffix: () => ChipSuffix
});
module.exports = __toCommonJS(ChipContent_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var styles = __toESM(require("./ChipContent.sty.js"));
var import_useChipContext = require("./contexts/useChipContext.js");
const ChipContent = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    children,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    $isPrefixWithKey,
    ...remainingProps
  } = props;
  const { size } = (0, import_useChipContext.useChipContext)();
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      ref: forwardedRef,
      className: (0, import_clsx.default)(
        styles.content({
          size
        }),
        consumerClassName
      ),
      style: consumerStyle,
      "data-testid": dataTestId,
      ...remainingProps,
      children
    }
  );
});
const ChipPrefix = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    ...restProps
  } = props;
  const { hasKey } = (0, import_useChipContext.useChipContext)();
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    ChipContent,
    {
      "data-testid": dataTestId,
      ref: forwardedRef,
      className: (0, import_clsx.default)(
        {
          [styles.prefix]: hasKey
        },
        consumerClassName
      ),
      style: consumerStyle,
      $isPrefixWithKey: hasKey,
      ...restProps
    }
  );
});
const ChipSuffix = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    ...restProps
  } = props;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    ChipContent,
    {
      "data-testid": dataTestId,
      ref: forwardedRef,
      className: consumerClassName,
      style: consumerStyle,
      ...restProps
    }
  );
});
ChipPrefix.displayName = "Chip.Prefix";
ChipSuffix.displayName = "Chip.Suffix";
