// @dynatrace/strato-components/content/chip/ChipDeleteButton.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ChipDeleteButton_exports = {};
__export(ChipDeleteButton_exports, {
  ChipDeleteButton: () => ChipDeleteButton
});
module.exports = __toCommonJS(ChipDeleteButton_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var styles = __toESM(require("./ChipDeleteButton.sty.js"));
var import_useChipContext = require("./contexts/useChipContext.js");
var import_Button = require("../../buttons/button/Button.js");
const ChipDeleteButton = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const intl = (0, import_react_intl.useIntl)();
  const deleteChip = intl.formatMessage({
    defaultMessage: "Delete chip",
    id: "JfsC0aOiAl7G3esV",
    description: "Aria label used to describe the delete chip button."
  });
  const {
    disabled: propDisabled,
    "aria-label": ariaLabelProp = deleteChip,
    "aria-describedby": ariaDescribedByProp,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    ...remainingProps
  } = props;
  const {
    disabled,
    variant,
    color,
    size = "default",
    valueId
  } = (0, import_useChipContext.useChipContext)();
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Button.Button,
    {
      ref: forwardedRef,
      "data-testid": dataTestId,
      variant,
      color,
      disabled: propDisabled ?? disabled,
      "aria-label": ariaLabelProp,
      "aria-describedby": ariaDescribedByProp || valueId,
      className: (0, import_clsx.default)(styles.base, styles.size[size], consumerClassName, {
        [styles.neutralEmphasizedOverride]: color === "neutral" && variant === "emphasized",
        [styles.disabledOverride]: propDisabled ?? disabled
      }),
      style: consumerStyle,
      ...remainingProps,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.XmarkIcon, { size: size === "default" ? size : "small" }) })
    }
  );
});
ChipDeleteButton.displayName = "Chip.DeleteButton";
