// @dynatrace/strato-components/content/chip/GroupedChip.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var GroupedChip_exports = {};
__export(GroupedChip_exports, {
  GroupedChip: () => GroupedChip
});
module.exports = __toCommonJS(GroupedChip_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_BaseChip = require("./BaseChip.js");
var styles = __toESM(require("./GroupedChip.sty.js"));
var import_useContainerSize = require("../../core/hooks/useContainerSize.js");
var import_useId = require("../../core/hooks/useId.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_useChipGroupContext = require("../chip-group/hooks/useChipGroupContext.js");
var import_useInternalChipGroupContext = require("../chip-group/hooks/useInternalChipGroupContext.js");
var import_chip_group_state_reducer = require("../chip-group/state/chip-group-state-reducer.js");
const GroupedChip = /* @__PURE__ */ (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const {
      children,
      size: propSize,
      disabled: propDisabled,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...remainingProps
    } = props;
    const uniqueId = (0, import_useId.useId)();
    const chipRef = (0, import_react.useRef)(null);
    const { width } = (0, import_useContainerSize.useContainerSize)(chipRef);
    const rootRef = (0, import_useMergeRefs.useMergeRefs)([chipRef, forwardedRef]);
    const { visibleKeys } = (0, import_useChipGroupContext.useChipGroupContext)();
    const internalChipGroupContext = (0, import_useInternalChipGroupContext.useInternalChipGroupContext)();
    const {
      disabled: chipGroupDisabled,
      size: chipGroupSize,
      dispatch
    } = internalChipGroupContext || {};
    const disabled = propDisabled ?? chipGroupDisabled ?? false;
    const size = propSize ?? chipGroupSize ?? "default";
    (0, import_react.useEffect)(() => {
      dispatch?.({
        type: import_chip_group_state_reducer.ChipGroupActions.ADDCHIP,
        payload: { uniqueId, width }
      });
      return () => {
        dispatch?.({
          type: import_chip_group_state_reducer.ChipGroupActions.REMOVECHIP,
          payload: {
            uniqueId
          }
        });
      };
    }, []);
    (0, import_react.useEffect)(() => {
      dispatch?.({
        type: import_chip_group_state_reducer.ChipGroupActions.UPDATECHIP,
        payload: { uniqueId, width: width === 0 ? void 0 : width }
      });
    }, [dispatch, uniqueId, width]);
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "div",
      {
        ref: rootRef,
        className: (0, import_clsx.default)("chip", styles.grouped, {
          [styles.first]: visibleKeys.includes(uniqueId) && visibleKeys.length === 1,
          [styles.hidden]: !visibleKeys.includes(uniqueId)
        }),
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_BaseChip.BaseChip,
          {
            ...remainingProps,
            disabled,
            size,
            children
          }
        )
      }
    );
  }
);
