// @dynatrace/strato-components/esm/editors/code-editor/auto-completion/javascript.js
import {
  completeFromList,
  ifNotIn
} from "@codemirror/autocomplete";
import {
  localCompletionSource,
  scopeCompletionSource,
  snippets
} from "@codemirror/lang-javascript";
const keywords = "break case const continue default delete export extends false finally in instanceof let new return static super switch this throw true typeof var yield".split(" ").map((kw) => ({
  label: kw,
  type: "keyword"
}));
const dontComplete = [
  "TemplateString",
  "String",
  "RegExp",
  "LineComment",
  "BlockComment",
  "VariableDefinition",
  "TypeDefinition",
  "Label",
  "PropertyDefinition",
  "PropertyName",
  "PrivatePropertyDefinition",
  "PrivatePropertyName"
];
const javascriptAutocompleteProvider = async (context) => {
  const localCompletionResult = localCompletionSource(context);
  const javascriptCompletionResult = await javascriptCompletionSource()(
    context
  );
  const scopeCompletionResult = await scopeCompletionSource(globalThis)(
    context
  );
  const before = context.matchBefore(/\w+/);
  const completions = extractCompletions([
    localCompletionResult,
    javascriptCompletionResult,
    scopeCompletionResult
  ]);
  if (completions.length === 0) {
    return { from: context.pos, suggestions: [] };
  }
  return {
    from: before ? before.from : context.pos,
    keepOrder: true,
    validFor: /\w+/,
    suggestions: transformSuggestions(completions)
  };
};
const javascriptCompletionSource = () => {
  return ifNotIn(dontComplete, completeFromList(snippets.concat(keywords)));
};
const extractCompletions = (completionResultList) => {
  let completions = [];
  for (const completionResult of completionResultList) {
    if (completionResult !== null) {
      completions = completions.concat(completionResult.options);
    }
  }
  return completions;
};
const transformSuggestions = (completions) => {
  const suggestions = [];
  for (const completion of completions) {
    suggestions.push({
      postfix: completion.detail,
      label: completion.label,
      type: completion.type,
      apply: completion.apply
    });
  }
  return suggestions;
};
export {
  javascriptAutocompleteProvider
};
