// @dynatrace/strato-components/tables/DataTable/features/Layout/layout-utils.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var layout_utils_exports = {};
__export(layout_utils_exports, {
  getColumnTemplateEntry: () => getColumnTemplateEntry,
  getPrefixColumnWidth: () => getPrefixColumnWidth,
  getTargetTemplate: () => getTargetTemplate,
  processBuiltinColumn: () => processBuiltinColumn
});
module.exports = __toCommonJS(layout_utils_exports);
var import_width_utils = require("./width-utils.js");
var import_logging = require("../../../../core/utils/logging.js");
var import_constants = require("../../constants.js");
var import_RowSelection = require("../RowSelection/RowSelection.js");
function getPrefixColumnWidth(column, rowDensity, rowThresholds) {
  const isFirstColumn = column.getIsFirstColumn();
  const iconPadding = (0, import_RowSelection.isRowSelectionColumnDef)(column.columnDef) ? 0 : 2 * import_constants.DATA_TABLE_DEFAULT_ICON_PADDING;
  let rowDensityPadding = import_constants.DATA_TABLE_CELL_VERTICAL_PADDING_MAP[rowDensity];
  let thresholdPillSize = 0;
  if (isFirstColumn) {
    rowDensityPadding *= 2;
    const hasThresholdPill = rowThresholds?.some(
      (threshold) => threshold.type === "pill"
    );
    thresholdPillSize = hasThresholdPill ? import_constants.DATA_TABLE_THRESHOLD_PILL_SIZE : 0;
  }
  return import_constants.DATA_TABLE_DEFAULT_ICON_SIZE + iconPadding + thresholdPillSize + rowDensityPadding;
}
function getTargetTemplate(pinnedState, templates) {
  if (!pinnedState) {
    return templates.center;
  }
  return templates[pinnedState];
}
function clampWidthToPx(width, minWidth, maxWidth) {
  return `${Math.min(Math.max(width, minWidth), maxWidth)}px`;
}
function calculateParentDependentWidth(column, parentWidth, getMeasuredWidth) {
  let totalDefinedChildWidths = 0;
  let undefinedWidthChildCount = 1;
  column.parent?.getVisibleLeafColumns().forEach((childCol) => {
    const childWidth = childCol.columnDef.width;
    if (childWidth === void 0 && childCol.id !== column.id) {
      undefinedWidthChildCount++;
    } else if (typeof childWidth === "number") {
      totalDefinedChildWidths += childWidth;
    } else if (childWidth === "auto" || childWidth === "content") {
      totalDefinedChildWidths += getMeasuredWidth(childCol.id) ?? 0;
    } else if ((0, import_width_utils.isFractionWidth)(childWidth)) {
      (0, import_logging.devLog)(
        "Fraction width is not allowed in combination with a fixed width parent column."
      );
    }
  });
  const remainingWidth = (parentWidth - totalDefinedChildWidths) / undefinedWidthChildCount;
  return `${remainingWidth}px`;
}
function getFractionWidthTemplateValue(parent, width, minWidth) {
  if (!parent || !(0, import_width_utils.isFractionWidth)(parent.columnDef.width)) {
    return `minmax(${minWidth}px, ${width})`;
  }
  const parentFraction = Number.parseFloat(parent.columnDef.width);
  const columnFraction = Number.parseFloat(width);
  const visibleSiblingColumns = parent.getVisibleLeafColumns();
  const siblingColumnsSum = visibleSiblingColumns.reduce((sum, col) => {
    return (0, import_width_utils.isFractionWidth)(col.columnDef.width) ? sum + Number.parseFloat(col.columnDef.width) : sum;
  }, 0);
  const calculatedFraction = parentFraction * (columnFraction / siblingColumnsSum);
  return `${calculatedFraction}fr`;
}
function getContentOrAutoWidthTemplateValue(measuredWidth, isContent) {
  if (measuredWidth !== void 0) {
    return `${measuredWidth}px`;
  }
  return isContent ? "max-content" : "minmax(min-content, auto)";
}
function getStringWidthTemplateValue(column, width, minWidth, getMeasuredWidth) {
  if ((0, import_width_utils.isFractionWidth)(width)) {
    return getFractionWidthTemplateValue(column.parent, width, minWidth);
  }
  if ((0, import_width_utils.isWidthExtendedConfig)(width)) {
    return getContentOrAutoWidthTemplateValue(
      getMeasuredWidth(column.id),
      width.type === "content"
    );
  }
  return getContentOrAutoWidthTemplateValue(
    getMeasuredWidth(column.id),
    width === "content"
  );
}
function getUndefinedWidthTemplateValue(column, minWidth, getMeasuredWidth) {
  const parentWidth = column.parent?.columnDef.width;
  if (column.parent && typeof parentWidth === "number") {
    return calculateParentDependentWidth(column, parentWidth, getMeasuredWidth);
  }
  const measuredWidth = getMeasuredWidth(column.id, true);
  return measuredWidth === void 0 ? `minmax(${minWidth}px, 1fr)` : `${measuredWidth}px`;
}
function processBuiltinColumn(column, rowDensity, rowThresholds, columnGridTemplates) {
  if (column.columnDef.meta?.builtinPlacement === "prefix") {
    columnGridTemplates.left.push(
      `${getPrefixColumnWidth(column, rowDensity, rowThresholds)}px`
    );
    return true;
  }
  if (column.columnDef.meta?.spacer === true) {
    columnGridTemplates.center.push("auto");
    return true;
  }
  if (column.columnDef.meta?.builtinPlacement === "suffix") {
    columnGridTemplates.right.push("min-content");
    return true;
  }
  return false;
}
function getColumnTemplateEntry(column, columnSizing, columnSizingInfo, getMeasuredWidth) {
  const {
    minWidth = import_constants.DATA_TABLE_DEFAULT_MIN_COLUMN_WIDTH,
    width,
    maxWidth = Number.MAX_SAFE_INTEGER
  } = column.columnDef;
  if (columnSizingInfo.isResizingColumn === column.id) {
    return `var(${import_constants.DATA_TABLE_STRATO_RESIZING_SIZE_CSS_VAR})`;
  }
  if (columnSizing[column.id] !== void 0) {
    return clampWidthToPx(columnSizing[column.id], minWidth, maxWidth);
  }
  if (width === void 0) {
    return getUndefinedWidthTemplateValue(column, minWidth, getMeasuredWidth);
  }
  if (typeof width === "number") {
    return clampWidthToPx(width, minWidth, maxWidth);
  }
  if (typeof width === "string" || (0, import_width_utils.isWidthExtendedConfig)(width)) {
    return getStringWidthTemplateValue(
      column,
      width,
      minWidth,
      getMeasuredWidth
    );
  }
  return void 0;
}
