// @dynatrace/strato-components/tables/DataTable/features/Layout/Layout.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Layout_exports = {};
__export(Layout_exports, {
  Layout: () => Layout,
  SPACER_COLUMN: () => SPACER_COLUMN,
  useLayout: () => useLayout
});
module.exports = __toCommonJS(Layout_exports);
var import_layout_utils = require("./layout-utils.js");
var import_useDataTableConfig = require("../../../hooks/useDataTableConfig.js");
var import_constants = require("../../constants.js");
var import_utils = require("../../hooks/useTable/utils.js");
var import_src = require("../../tanstack-table-core-override/src/index.js");
const SPACER_COLUMN = {
  id: import_constants.DATA_TABLE_SPACER_COLUMN_ID,
  header: () => void 0,
  cell: () => void 0,
  enableSorting: false,
  enablePinning: false,
  enableGrouping: false,
  enableUserActions: false,
  enableResizing: false,
  enableHiding: false,
  meta: {
    builtin: true,
    builtinPlacement: "suffix",
    hasCustomCellRenderer: true,
    hasCustomHeaderRenderer: true,
    spacer: true,
    cellDeps: () => {
      return [];
    }
  }
};
const Layout = {
  getInitialState(state) {
    return {
      columnMeasureInfo: {},
      observedHeaderHeight: 0,
      ...state
    };
  },
  createHeader(header, table) {
    header.getGridPosition = (0, import_src.memo)(
      (headers) => [headers],
      (headers) => {
        let gridColumnStart = 0;
        for (const headerInGroup of headers ?? []) {
          if (headerInGroup === header) {
            break;
          }
          gridColumnStart = gridColumnStart + headerInGroup.colSpan;
        }
        const colSpan = header.colSpan;
        const gridRowStart = header.column.depth;
        const gridRowEnd = header.depth;
        return {
          gridColumnStart: gridColumnStart + 1,
          gridColumnEnd: `span ${colSpan}`,
          gridRowStart: gridRowStart + 1,
          gridRowEnd: `span ${gridRowEnd - gridRowStart}`
        };
      },
      (0, import_utils.getMemoOptions)(table.options, "debugHeaders", "header.getGridPosition")
    );
  },
  createColumn(column, table) {
    column.getGridColumnPosition = (columnIndex) => ({
      // Offset the index by one as the CSS grid column is a one based position.
      gridColumnStart: (columnIndex ?? 0) + 1,
      gridColumnEnd: "span 1"
    });
  },
  createTable(table) {
    (0, import_utils.assignRowMethods)(table, {
      getGridRowPosition(rowIndex = -1) {
        return {
          // Offset the index by one as the CSS grid column is a one based position.
          gridColumnStart: 1,
          // The row should span all columns.
          gridColumnEnd: "row-end",
          // The CSS grid column is a one based position, so we need to offset the index by one.
          // To account for the virtualized rows at the start of the grid, a spacer row is inserted.
          // Therefore we need to offset the index by two in total
          gridRowStart: rowIndex + 2,
          // The row should span only one line from the start edge.
          gridRowEnd: "span 1"
        };
      }
    });
    table.getHasFractionalTemplate = () => {
      const template = table.getGridTemplateColumns();
      return template.includes("fr");
    };
    table.getVisibleHeaderGroupsCount = () => {
      if (table.getVisualVariantOptions().headers === "hidden") {
        return 0;
      }
      return table.getHeaderGroups().length;
    };
    table.setObservedHeaderHeight = (height) => {
      if (height !== table.getState().observedHeaderHeight) {
        table.setState((state) => ({
          ...state,
          observedHeaderHeight: height
        }));
      }
    };
    table.getHeaderSize = (headerId) => {
      const escapedId = CSS.escape(headerId);
      const headerSize = table.getRowVirtualizationContainer()?.querySelector(
        `[role="columnheader"][data-headerid=${escapedId}], [role="columnheader"][data-columnid=${escapedId}], [role="presentation"][data-columnid=${escapedId}]`
      )?.getBoundingClientRect();
      if (headerSize === void 0) {
        return {
          width: import_constants.DATA_TABLE_DEFAULT_COLUMN_WIDTH,
          height: table.estimateRowHeight({ isHeaderRow: true })
        };
      }
      return {
        width: headerSize.width,
        height: headerSize.height
      };
    };
    table.getColumnOrHeaderWidth = (columnOrHeaderId) => {
      const hasHiddenHeaders = table.getVisualVariantOptions().headers === "hidden";
      const storedWidth = table.getMeasuredColumnWidth(columnOrHeaderId);
      if (!hasHiddenHeaders && storedWidth !== void 0) {
        return storedWidth;
      }
      const escapedId = CSS.escape(columnOrHeaderId);
      const selectorString = hasHiddenHeaders ? `[role="cell"][data-columnid="${escapedId}"]` : `[role="columnheader"][data-headerid="${escapedId}"], [role="columnheader"][data-columnid="${escapedId}"], [role="presentation"][data-columnid="${escapedId}"]`;
      const columnWidth = table.getRowVirtualizationContainer()?.querySelector(selectorString)?.getBoundingClientRect().width;
      return columnWidth ?? import_constants.DATA_TABLE_DEFAULT_COLUMN_WIDTH;
    };
    table.getTotalHeaderHeight = (0, import_src.memo)(
      () => [
        table.getVisualVariantOptions(),
        table.getVisibleLeafColumnsWithoutBuiltin(),
        table.getHeaderGroups(),
        table.getState().observedHeaderHeight
      ],
      (visualVariantOptions, visibleLeafColumns, tableHeaderGroups, observedHeaderHeight) => {
        if (visualVariantOptions.headers === "hidden" || visibleLeafColumns.length === 0 || tableHeaderGroups.length === 1 && tableHeaderGroups.at(0)?.headers.length === 0) {
          return 0;
        }
        if (observedHeaderHeight > 0) {
          return observedHeaderHeight;
        }
        const bottomBorderOffset = import_constants.DATA_TABLE_ROW_BORDER;
        const fallbackHeight = table.estimateRowHeight({ isHeaderRow: true }) * table.getVisibleHeaderGroupsCount() + bottomBorderOffset;
        return fallbackHeight;
      },
      (0, import_utils.getMemoOptions)(table.options, "debugTable", "table.getTotalHeaderHeight")
    );
    table.estimateRowHeight = (0, import_src.memo)(
      (props) => [
        props?.index,
        props?.isHeaderRow,
        table.getVisualVariantOptions()
      ],
      (index, isHeaderRow, { rowDensity, rowSeparation }) => {
        const rowHeight = import_constants.DATA_TABLE_ROW_HEIGHT_MAP[rowDensity];
        const paginationFlatRows = table.getPaginationRowModel().flatRows;
        const isLastRowOnPage = index === paginationFlatRows.length - 1;
        const addBorder = index !== void 0 || isHeaderRow === true;
        if (addBorder === false || rowSeparation === "none" || isLastRowOnPage) {
          return rowHeight;
        }
        return rowHeight + import_constants.DATA_TABLE_ROW_BORDER;
      },
      (0, import_utils.getMemoOptions)(table.options, "debugTable", "table._estimateRowHeight")
    );
    table.getHeaderGridTemplateRows = () => {
      switch (table.getVisibleHeaderGroupsCount()) {
        case 0:
          return "[body-start]";
        case 1:
          return "[header-start] max-content [header-end body-start]";
        default:
          return "[header-start] max-content max-content [header-end body-start]";
      }
    };
    table.getGridTemplateRows = (0, import_src.memo)(
      () => {
        const { getVirtualItems, range } = table.getRowVirtualizer() ?? {};
        return [
          getVirtualItems?.(),
          range?.startIndex,
          range?.endIndex,
          table.getVisibleHeaderGroupsCount(),
          table.options.data.length
        ];
      },
      (virtualItems) => {
        const measurementsCache = table.getRowVirtualizer()?.measurementsCache;
        const headerRows = table.getHeaderGridTemplateRows();
        if (!measurementsCache || !virtualItems?.length) {
          return headerRows;
        }
        const bodyRows = virtualItems.map((virtualItem) => {
          const rowHeight = virtualItem.size ?? table.estimateRowHeight({ index: virtualItem.index });
          return `minmax(min-content, ${rowHeight}px)`;
        }).join(" ");
        const firstVirtualItem = virtualItems[0];
        let spacerRowHeight = 0;
        if (firstVirtualItem.index > 0) {
          const scrollPaddingStart = table.getTotalHeaderHeight();
          spacerRowHeight = firstVirtualItem.start - scrollPaddingStart;
        }
        return `${headerRows} ${spacerRowHeight}px ${bodyRows} [body-end]`;
      },
      (0, import_utils.getMemoOptions)(table.options, "debugTable", "table.getGridTemplateRows")
    );
    table.getGridTemplateColumns = (0, import_src.memo)(
      () => {
        const state = table.getState();
        return [
          state.columnSizing,
          state.columnSizingInfo,
          state.columnPinning,
          state.columnOrder,
          state.columnVisibility,
          table.getSubRowIndicatorColumnId(),
          JSON.stringify([...table.getMeasuredColumnWidths()])
        ];
      },
      (columnSizing, columnSizingInfo) => {
        const leftVisibleColumns = table.getLeftVisibleLeafColumns();
        const centerVisibleColumns = table.getCenterVisibleLeafColumns();
        const rightVisibleColumns = table.getRightVisibleLeafColumns();
        if (leftVisibleColumns === void 0 || centerVisibleColumns === void 0 || rightVisibleColumns === void 0 || columnSizing === void 0 || columnSizingInfo === void 0) {
          return "";
        }
        function getMeasuredWidth(columnId, isFractionWidth = false) {
          if (table.getRowModel().rows.length === 0) {
            const headerCell = table.getColumnVirtualizer()?.scrollElement?.querySelector(
              `[role="columnheader"][data-columnid="${CSS.escape(
                columnId
              )}"]`
            );
            const headerWidth = headerCell?.getBoundingClientRect().width;
            if (headerWidth === void 0) {
              return import_constants.DATA_TABLE_MIN_WIDTH;
            }
            return isFractionWidth ? headerWidth : Math.ceil(headerWidth * 10) / 10;
          }
          return table.getMeasuredColumnWidth(columnId);
        }
        const columnGridTemplates = {
          left: [],
          center: [],
          right: []
        };
        const orderedVisibleColumns = [
          ...leftVisibleColumns,
          ...centerVisibleColumns,
          ...rightVisibleColumns
        ];
        for (const column of orderedVisibleColumns) {
          const isBuiltinColumn = (0, import_layout_utils.processBuiltinColumn)(
            column,
            table.getVisualVariantOptions().rowDensity,
            table.options.rowThresholds,
            columnGridTemplates
          );
          if (isBuiltinColumn) {
            continue;
          }
          const entry = (0, import_layout_utils.getColumnTemplateEntry)(
            column,
            columnSizing,
            columnSizingInfo,
            getMeasuredWidth
          );
          if (entry !== void 0) {
            const targetTemplate = (0, import_layout_utils.getTargetTemplate)(
              column.getIsPinned(),
              columnGridTemplates
            );
            targetTemplate.push(entry);
          }
        }
        const combinedGridTemplates = [
          ...columnGridTemplates.left,
          ...columnGridTemplates.center,
          ...columnGridTemplates.right
        ];
        return `${combinedGridTemplates.join(" ")} [row-end]`;
      },
      (0, import_utils.getMemoOptions)(
        table.options,
        "debugTable",
        "table.getGridTemplateColumns"
      )
    );
  }
};
function useLayout(props, options) {
  const configFullWidth = (0, import_useDataTableConfig.useDataTableConfigElement)("fullWidth");
  const configFullHeight = (0, import_useDataTableConfig.useDataTableConfigElement)("fullHeight");
  options.layout = {
    ...options.layout,
    fullWidth: props.fullWidth ?? configFullWidth,
    fullHeight: props.fullHeight ?? configFullHeight
  };
}
