// @dynatrace/strato-components/esm/charts/index.js
import { InteractionsOverlay } from "./core/components/interactions-overlay/InteractionsOverlay.js";
import { AnnotationsChartConfig } from "./annotations-chart/contexts/AnnotationsChartConfig.context.js";
import { AnnotationsChart } from "./annotations-chart/AnnotationsChart.js";
import { CategoricalBarChart } from "./categorical-bar/CategoricalBarChartRoot.js";
import { CategoricalBarChartConfig } from "./categorical-bar/hooks/useCategoricalBarChartConfig.js";
import { AnnotationsAtom } from "./core/components/annotations/AnnotationsAtom.js";
import { MarkerAtom } from "./core/components/annotations/slots/MarkerAtom.js";
import { TrackAtom } from "./core/components/annotations/slots/TrackAtom.js";
import {
  HORIZONTAL_LABEL_LAYOUT_TYPE,
  VERTICAL_LABEL_LAYOUT_TYPE
} from "./core/components/axes/types.js";
import { ChartGrid } from "./core/components/chart-grid/ChartGridRoot.js";
import { ChartLayout } from "./core/components/chart-layout/ChartLayout.js";
import { getLegendOrientation } from "./core/components/chart-layout/utils/orientation.js";
import { EmptyState } from "./core/components/chart-states/EmptyState.js";
import { ErrorState } from "./core/components/chart-states/ErrorState.js";
import { ChartLegend } from "./core/components/legend/categorical/ChartLegend.js";
import { RangeLegend } from "./core/components/legend/range/RangeLegend.js";
import { LegendSizeContext } from "./core/components/legend/shared/context/LegendSizeContext.js";
import { ChartSeriesAction } from "./core/components/series-action/ChartSeriesAction.js";
import { useChartActionsMenu } from "./core/components/series-action/hooks/useChartActionsMenu.js";
import { ThresholdBar } from "./core/components/threshold-bar/ThresholdBar.js";
import { ThresholdIndicator } from "./core/components/threshold-bar/indicators/ThresholdIndicator.js";
import { ChartToolbarRenderer } from "./core/components/toolbar/ChartToolbarRenderer.js";
import { ChartTooltip } from "./core/components/tooltip/ChartTooltip/ChartTooltip.js";
import { ChartTooltipSlots } from "./core/components/tooltip/types/ChartTooltip.config.js";
import { ChartTooltipAtoms } from "./core/components/tooltip/types/ChartTooltip.config.js";
import { ChartTooltipTemplate } from "./core/components/tooltip/types/ChartTooltip.config.js";
import { GeoItem } from "./core/components/tooltip/ChartTooltip/components/GeoItem.js";
import { SeriesItem } from "./core/components/tooltip/ChartTooltip/components/SeriesItem/SeriesItem.js";
import { TimeFrameHeader } from "./core/components/tooltip/ChartTooltip/components/TimeFrameHeader/TimeFrameHeader.js";
import { EventItem } from "./core/components/tooltip/ChartTooltip/components/EventItem/EventItem.js";
import { KeyValue } from "./core/components/tooltip/ChartTooltip/components/KeyValue.js";
import { TextHeader } from "./core/components/tooltip/ChartTooltip/components/TextHeader.js";
import { DurationChip } from "./core/components/tooltip/ChartTooltip/components/DurationChip.js";
import { SharedChartInteractions } from "./core/components/shared-chart-interactions/SharedChartInteractions.js";
import { OverlayTooltip } from "./core/components/overlay-tooltip/OverlayTooltip.js";
import { ChartTooltip as ChartTooltip2 } from "./core/components/overlay-tooltip/OverlayTooltip.js";
import { GEO_DEFAULTS } from "./core/components/overlay-tooltip/defaults.js";
import { useDefaultConfig } from "./core/components/overlay-tooltip/hooks/useDefaultConfig.js";
import { useDefaultLabels } from "./core/components/overlay-tooltip/hooks/useDefaultLabels.js";
import {
  OverlayProvider,
  useOverlay
} from "./core/components/overlay-tooltip/contexts/OverlayContext.js";
import {
  TooltipProvider,
  useTooltip
} from "./core/components/overlay-tooltip/contexts/TooltipContext.js";
import { useTooltipReducer } from "./core/components/overlay-tooltip/useTooltipReducer.js";
import {
  useTooltipActions
} from "./core/components/overlay-tooltip/useTooltipActions.js";
import { useDefaultLabels as useDefaultLabels2 } from "./core/components/overlay-tooltip/hooks/useDefaultLabels.js";
import {
  OverlayTooltipHeaderStory,
  OverlayTooltipTruncationStory,
  OverlayTooltipWIPStory,
  OverlayTooltipCustomStory
} from "./core/components/overlay-tooltip/_to-be-removed/_OverlayTooltipWIPStory.js";
import {
  SIXTY_FPS,
  DEFAULT_TRUNCATION_MODE,
  CSS_DOWNLOAD_FILENAME
} from "./core/constants.js";
import { getTabIndexValue } from "./core/utils/a11yToolbarControls.js";
import { getTabIndexKey } from "./core/utils/a11yToolbarControls.js";
import { buildQuantizedRangeItems } from "./core/utils/build-quantized-range-items.js";
import { buildCategoricalItemsFromScale } from "./core/utils/build-categorical-items-from-scale.js";
import { buildLegendItemsFromRange } from "./core/utils/build-legend-items-from-range.js";
import { buildQuantizeScaleFromColorPalette } from "./core/utils/build-quantize-scale-from-color-palette.js";
import { buildThresholdScaleFromRanges } from "./core/utils/build-threshold-scale-from-ranges.js";
import { coerceSizeValue } from "./core/utils/coerceSizeValue.js";
import { isCustomColorPalette } from "./core/types/color-palette.js";
import { useA11yToolbarControls } from "./core/hooks/useA11yToolbarControls.js";
import { useA11yToolbarTabIndex } from "./core/hooks/useA11yToolbarTabIndex.js";
import { useThrottledResizeObserver } from "./core/hooks/useThrottledResizeObserver.js";
import { useObserveContainerSize } from "./core/hooks/useObserveContainerSize.js";
import { getCanvasColor } from "./core/components/canvas/utils/canvas-color.js";
import {
  pointsDisplayOptions
} from "./core/components/canvas/types/canvas-shape.js";
import { useGraphSize } from "./core/components/chart-layout/hooks/useGraphSize.js";
import { useChartLayoutOrientation } from "./core/components/chart-layout/hooks/useChartLayoutOrientation.js";
import { useChartLayoutLoading } from "./core/components/chart-layout/hooks/useChartLayoutLoading.js";
import { useAutoLegendRefresh } from "./core/components/chart-layout/hooks/useAutoLegendRefresh.js";
import { useGridPlotSize } from "./core/components/chart-grid/hooks/useGridPlotSize.js";
import { useChartLayoutSize } from "./core/components/chart-layout/hooks/useChartLayoutSize.js";
import { useLegendSize } from "./core/components/legend/shared/context/LegendSizeContext.js";
import { getTickFormatterMethod } from "./core/utils/get-tick-formatter-method.js";
import {
  sequentialColoringOptions,
  divergingColoringOptions,
  themedColoringOptions,
  automaticColoringOptions,
  colorPaletteOptions,
  colorPaletteModeOptions
} from "./core/types/color-palette.js";
import { ChartToolbar } from "./core/slots/toolbar/ChartToolbar.js";
import {
  HistogramChart,
  HistogramAnnotations
} from "./histogram/HistogramChart.js";
import { HistogramChartConfig } from "./histogram/contexts/HistogramChartConfig.context.js";
import { HoneycombChart } from "./honeycomb/HoneycombChart.js";
import { HoneycombChartConfig } from "./honeycomb/context/honeycomb-chart-config.context.js";
import { MultiMeterBarChart } from "./meter-bar/multi-meter-bar/MultiMeterBarChart.js";
import { MultiMeterBarChartConfig } from "./meter-bar/multi-meter-bar/context/MultiMeterBarChartConfigContext.js";
import {
  multiMeterBarChartTooltipVariantOptions
} from "./meter-bar/multi-meter-bar/types/multi-meter-bar-chart.js";
import { MeterBarChart } from "./meter-bar/single-meter-bar/MeterBarChart.js";
import { MeterBarChartThresholdLegend } from "./meter-bar/single-meter-bar/components/MeterBarChartThresholdLegend/MeterBarChartThresholdLegend.js";
import { MeterBarChartConfig } from "./meter-bar/single-meter-bar/context/MeterBarChartConfigContext.js";
import { PieChart } from "./pie/PieChart.js";
import { DonutChart } from "./pie/DonutChart.js";
import {
  DonutChartConfig,
  PieChartConfig
} from "./pie/context/ChartConfigContext.js";
import { SingleValue } from "./single-value/SingleValue.js";
import { SingleValueGrid } from "./single-value-grid/SingleValueGrid.js";
import { SingleValueConfig } from "./single-value/hooks/use-single-value-config.js";
import { Sparkline } from "./sparkline/Sparkline.js";
import {
  TimeseriesAnnotations,
  TimeseriesChart
} from "./timeseries/TimeseriesChart.js";
import { TimeseriesChartConfig } from "./timeseries/context/TimeseriesChartConfig.context.js";
import { ChartInteractions } from "./core/slots/chart-interactions/ChartInteractions.js";
import {
  convertQueryResultToTimeseries,
  convertToTimeseries
} from "./timeseries/utils/convert-to-timeseries/convert-to-timeseries.js";
import { convertToTimeseriesBand } from "./timeseries/utils/convert-to-timeseries/convert-to-timeseries-band.js";
import {
  getFieldsArrayForRecordIndex,
  getTypesForRecordIndex,
  convertToFieldEntries
} from "./timeseries/utils/convert-to-timeseries/utils/getFieldsArrayForRecordIndex.js";
import { checkTimeseriesFormat } from "./timeseries/utils/convert-to-timeseries/utils/check-timeseries-format.js";
import { getDimensions } from "./timeseries/utils/convert-to-timeseries/utils/get-dimensions.js";
import { _getMetadataMetricsUnitForFieldName } from "./timeseries/utils/convert-to-timeseries/utils/get-metadata-metrics-unit-for-field-name.js";
import { XYChart, XYChartAnnotations } from "./xy-chart/XYChartRoot.js";
import { XYChartConfig } from "./xy-chart/context/XYChartConfig.context.js";
import { XYChartTooltipTemplate } from "./xy-chart/components/tooltip/XYChartTooltipTemplate/XYChartTooltipTemplate.js";
import { XYChartTooltipHeader } from "./xy-chart/components/tooltip/XYChartTooltipHeader/XYChartTooltipHeader.js";
import { XYChartTooltipBody } from "./xy-chart/components/tooltip/XYChartTooltipBody/XYChartTooltipBody.js";
import { TreeMap } from "./tree-map/TreeMap.js";
import { TreeMapConfig } from "./tree-map/context/TreeMapConfigContext.js";
import { useMousePosition } from "./core/hooks/useMousePosition.js";
import { truncateText } from "./core/utils/text-truncation.js";
import { GaugeChart } from "./gauge/GaugeChart.js";
import { GaugeChartConfig } from "./gauge/contexts/GaugeChartConfigContext.js";
import { colorPaletteMap } from "./core/types/color-palettes-definitions.js";
import { automaticColoringMap } from "./core/types/color-palettes-definitions.js";
export {
  AnnotationsChart,
  AnnotationsChartConfig,
  CSS_DOWNLOAD_FILENAME,
  CategoricalBarChart,
  CategoricalBarChartConfig,
  ChartInteractions,
  ChartSeriesAction,
  ChartToolbar,
  ChartTooltip2 as ChartTooltip,
  DEFAULT_TRUNCATION_MODE,
  DonutChart,
  DonutChartConfig,
  GaugeChart,
  GaugeChartConfig,
  HORIZONTAL_LABEL_LAYOUT_TYPE,
  HistogramAnnotations,
  HistogramChart,
  HistogramChartConfig,
  HoneycombChart,
  HoneycombChartConfig,
  MeterBarChart,
  MeterBarChartConfig,
  MultiMeterBarChart,
  MultiMeterBarChartConfig,
  PieChart,
  PieChartConfig,
  SharedChartInteractions,
  SingleValue,
  SingleValueConfig,
  SingleValueGrid,
  Sparkline,
  TimeseriesAnnotations,
  TimeseriesChart,
  TimeseriesChartConfig,
  TreeMap,
  TreeMapConfig,
  VERTICAL_LABEL_LAYOUT_TYPE,
  XYChart,
  XYChartAnnotations,
  XYChartConfig,
  AnnotationsAtom as _AnnotationsAtom,
  ChartGrid as _ChartGrid,
  ChartLayout as _ChartLayout,
  ChartLegend as _ChartLegend,
  ChartToolbarRenderer as _ChartToolbarRenderer,
  ChartTooltip as _ChartTooltip,
  ChartTooltipAtoms as _ChartTooltipAtoms,
  ChartTooltipSlots as _ChartTooltipSlots,
  ChartTooltipTemplate as _ChartTooltipTemplate,
  DurationChip as _DurationChip,
  EmptyState as _EmptyState,
  ErrorState as _ErrorState,
  EventItem as _EventItem,
  SIXTY_FPS as _FPS,
  GEO_DEFAULTS as _GEO_DEFAULTS,
  GeoItem as _GeoItem,
  InteractionsOverlay as _InteractionsOverlay,
  KeyValue as _KeyValue,
  LegendSizeContext as _LegendSizeContext,
  MarkerAtom as _MarkerAtom,
  MeterBarChartThresholdLegend as _MeterBarChartThresholdLegend,
  OverlayProvider as _OverlayProvider,
  OverlayTooltip as _OverlayTooltip,
  OverlayTooltipCustomStory as _OverlayTooltipCustomStory,
  OverlayTooltipHeaderStory as _OverlayTooltipHeaderStory,
  TooltipProvider as _OverlayTooltipProvider,
  OverlayTooltipTruncationStory as _OverlayTooltipTruncationStory,
  OverlayTooltipWIPStory as _OverlayTooltipWIPStory,
  RangeLegend as _RangeLegend,
  SeriesItem as _SeriesItem,
  TextHeader as _TextHeader,
  ThresholdBar as _ThresholdBar,
  ThresholdIndicator as _ThresholdIndicator,
  TimeFrameHeader as _TimeFrameHeader,
  TrackAtom as _TrackAtom,
  XYChartTooltipBody as _XYChartTooltipBody,
  XYChartTooltipHeader as _XYChartTooltipHeader,
  XYChartTooltipTemplate as _XYChartTooltipTemplate,
  automaticColoringMap as _automaticColoringMap,
  automaticColoringOptions as _automaticColoringOptions,
  buildCategoricalItemsFromScale as _buildCategoricalItemsFromScale,
  buildLegendItemsFromRange as _buildLegendItemsFromRange,
  buildQuantizeScaleFromColorPalette as _buildQuantizeScaleFromColorPalette,
  buildQuantizedRangeItems as _buildQuantizedRangeItems,
  buildThresholdScaleFromRanges as _buildThresholdScaleFromRanges,
  coerceSizeValue as _coerceSizeValue,
  colorPaletteMap as _colorPaletteMap,
  divergingColoringOptions as _divergingColoringOptions,
  getCanvasColor as _getCanvasColor,
  getLegendOrientation as _getLegendOrientation,
  _getMetadataMetricsUnitForFieldName,
  getTabIndexKey as _getTabIndexKey,
  getTabIndexValue as _getTabIndexValue,
  getTickFormatterMethod as _getTickFormatterMethod,
  isCustomColorPalette as _isCustomColorPalette,
  sequentialColoringOptions as _sequentialColoringOptions,
  themedColoringOptions as _themedColoringOptions,
  truncateText as _truncateText,
  useA11yToolbarControls as _useA11yToolbarControls,
  useA11yToolbarTabIndex as _useA11yToolbarTabIndex,
  useAutoLegendRefresh as _useAutoLegendRefresh,
  useChartActionsMenu as _useChartActionsMenu,
  useChartLayoutLoading as _useChartLayoutLoading,
  useChartLayoutOrientation as _useChartLayoutOrientation,
  useChartLayoutSize as _useChartLayoutSize,
  useDefaultConfig as _useDefaultConfig,
  useDefaultLabels as _useDefaultLabels,
  useGraphSize as _useGraphSize,
  useGridPlotSize as _useGridPlotSize,
  useLegendSize as _useLegendSize,
  useMousePosition as _useMousePosition,
  useObserveContainerSize as _useObserveContainerSize,
  useOverlay as _useOverlayChart,
  useTooltipActions as _useOverlayTooltipActions,
  useDefaultLabels2 as _useOverlayTooltipDefaultLabels,
  useTooltipReducer as _useOverlayTooltipReducer,
  useTooltip as _useOverlayTooltipStore,
  useThrottledResizeObserver as _useThrottledResizeObserver,
  checkTimeseriesFormat,
  colorPaletteModeOptions,
  colorPaletteOptions,
  convertQueryResultToTimeseries,
  convertToFieldEntries,
  convertToTimeseries,
  convertToTimeseriesBand,
  getDimensions,
  getFieldsArrayForRecordIndex,
  getTypesForRecordIndex,
  multiMeterBarChartTooltipVariantOptions,
  pointsDisplayOptions
};
