// @dynatrace/strato-components/esm/filters/filter-field/utils/validation-utils/normalize-validator-map.js
import {
  isFilterFieldKeySuggestionConfig,
  isFilterFieldKeySuggestionGroupConfig,
  isFilterFieldRichValuePredicate,
  isFilterFieldValidatorMapDurationSpecification,
  isFilterFieldValidatorMapTypeSpecification,
  isFilterFieldValueSuggestionGroupConfig
} from "./type-guards.js";
import { resolveGroupId } from "../suggestions-utils/suggestion-group-utils.js";
const KEY_GROUP_ID = "strato-filter-field-key-suggestions";
function toInternalValidatorMap(validatorMap) {
  if (!validatorMap) {
    return void 0;
  }
  const byKey = /* @__PURE__ */ new Map();
  const valuesMap = /* @__PURE__ */ new Map();
  const functionsMap = /* @__PURE__ */ new Map();
  const fallbackKeysList = [];
  let groups;
  const { keyPredicates, exhaustive, comparisonOperators } = validatorMap;
  if (keyPredicates !== void 0) {
    if (Array.isArray(keyPredicates)) {
      groups = processKeyPredicatesArray(
        keyPredicates,
        byKey,
        valuesMap,
        functionsMap,
        fallbackKeysList
      );
    } else {
      processKeyPredicatesObject(
        keyPredicates,
        byKey,
        valuesMap,
        functionsMap,
        fallbackKeysList
      );
    }
  }
  const filterStatementKeysMap = valuesMap.size > 0 || functionsMap.size > 0 ? { values: valuesMap, functions: functionsMap } : void 0;
  const resolvedGroups = groups && groups.length > 0 ? groups : void 0;
  const keyToGroup = /* @__PURE__ */ new Map();
  if (resolvedGroups) {
    for (const group of resolvedGroups) {
      for (const key of group.keys) {
        keyToGroup.set(key, group);
      }
    }
  } else {
    const implicitGroup = {
      id: `${KEY_GROUP_ID}-implicit-0`,
      keys: [...byKey.keys()]
    };
    for (const key of byKey.keys()) {
      keyToGroup.set(key, implicitGroup);
    }
  }
  return {
    normalizedKeys: {
      byKey,
      groups: resolvedGroups,
      keyToGroup
    },
    exhaustive,
    comparisonOperators,
    sortSuggestions: validatorMap.sortSuggestions,
    fallbackKeys: fallbackKeysList.length > 0 ? fallbackKeysList : void 0,
    filterStatementKeysMap
  };
}
function processKeyPredicatesArray(keyPredicates, byKey, valuesMap, functionsMap, fallbackKeysList) {
  const groups = [];
  let pendingKeys = [];
  function processPendingKeys() {
    if (pendingKeys.length === 0) {
      return;
    }
    groups.push({
      id: `${KEY_GROUP_ID}-implicit-${groups.length}`,
      keys: pendingKeys
    });
    pendingKeys = [];
  }
  for (let index = 0; index < keyPredicates.length; index++) {
    const item = keyPredicates[index];
    if (typeof item === "string") {
      byKey.set(item, {});
      pendingKeys.push(item);
      continue;
    }
    if (isFilterFieldKeySuggestionConfig(item)) {
      byKey.set(item.key, item);
      collectFallbackAndStatements(
        item.key,
        item,
        fallbackKeysList,
        valuesMap,
        functionsMap
      );
      pendingKeys.push(item.key);
      continue;
    }
    if (isFilterFieldKeySuggestionGroupConfig(item)) {
      processPendingKeys();
      const { label, "aria-label": ariaLabel } = item;
      const groupId = resolveGroupId(item, `key-group-${index}`);
      const groupKeys = [];
      for (const suggestion of item.suggestions) {
        if (typeof suggestion === "string") {
          byKey.set(suggestion, {});
          groupKeys.push(suggestion);
        } else if (isFilterFieldKeySuggestionConfig(suggestion)) {
          byKey.set(suggestion.key, suggestion);
          groupKeys.push(suggestion.key);
          collectFallbackAndStatements(
            suggestion.key,
            suggestion,
            fallbackKeysList,
            valuesMap,
            functionsMap
          );
        }
      }
      if (groupKeys.length > 0) {
        groups.push({
          id: groupId,
          label,
          "aria-label": ariaLabel,
          keys: groupKeys
        });
      }
    }
  }
  processPendingKeys();
  return groups;
}
function processKeyPredicatesObject(keyPredicates, byKey, valuesMap, functionsMap, fallbackKeysList) {
  for (const [key, rawPredicate] of Object.entries(keyPredicates)) {
    let config;
    if (isFilterFieldValidatorMapTypeSpecification(rawPredicate)) {
      config = { valuePredicate: rawPredicate };
    } else {
      config = {
        comparisonOperators: rawPredicate.comparisonOperators,
        valuePredicate: rawPredicate.valuePredicate,
        fallbackKey: rawPredicate.fallbackKey,
        suggestStatementOnValueMatch: rawPredicate.suggestStatementOnValueMatch,
        displayValue: rawPredicate.displayValue,
        details: rawPredicate.details
      };
    }
    byKey.set(key, config);
    collectFallbackAndStatements(
      key,
      config,
      fallbackKeysList,
      valuesMap,
      functionsMap
    );
  }
}
function collectFallbackAndStatements(key, config, fallbackKeysList, valuesMap, functionsMap) {
  if (config.fallbackKey) {
    fallbackKeysList.push({
      value: key,
      comparisonOperators: Array.isArray(config.fallbackKey) ? config.fallbackKey : void 0
    });
  }
  if (config.suggestStatementOnValueMatch) {
    if (typeof config.suggestStatementOnValueMatch === "function") {
      functionsMap.set(key, config.suggestStatementOnValueMatch);
    } else if (config.valuePredicate !== void 0 && typeof config.valuePredicate !== "function") {
      extractLiteralValuesToMap(config.valuePredicate, key, valuesMap);
    }
  }
}
function extractLiteralValuesToMap(valuePredicate, key, valuesMap) {
  if (Array.isArray(valuePredicate)) {
    for (const item of valuePredicate) {
      if (isFilterFieldValueSuggestionGroupConfig(item)) {
        for (const suggestion of item.suggestions) {
          if (!isFilterFieldValidatorMapTypeSpecification(suggestion)) {
            addToValuesMap(valuesMap, key, suggestion);
          }
        }
      } else if (!isFilterFieldValidatorMapTypeSpecification(item)) {
        addToValuesMap(valuesMap, key, item);
      }
    }
  } else if (isFilterFieldValueSuggestionGroupConfig(valuePredicate)) {
    for (const suggestion of valuePredicate.suggestions) {
      if (!isFilterFieldValidatorMapTypeSpecification(suggestion)) {
        addToValuesMap(valuesMap, key, suggestion);
      }
    }
  } else if (!isFilterFieldValidatorMapTypeSpecification(valuePredicate)) {
    addToValuesMap(valuesMap, key, valuePredicate);
  }
}
function addToValuesMap(valuesMap, key, value) {
  let stringValue;
  if (isFilterFieldValidatorMapDurationSpecification(value)) {
    stringValue = `${value.value}${value.unit}`;
  } else if (isFilterFieldRichValuePredicate(value)) {
    stringValue = `${value.value}`;
  } else {
    stringValue = `${value}`;
  }
  const existingKeys = valuesMap.get(stringValue) ?? [];
  if (!existingKeys.includes(key)) {
    valuesMap.set(stringValue, [...existingKeys, key]);
  }
}
export {
  KEY_GROUP_ID,
  toInternalValidatorMap
};
