// @dynatrace/strato-components/esm/filters/filter-field/utils/validation-utils/operator-specification.js
const OperatorsWithNames = {
  "=": "equals",
  "!=": "not-equals",
  "<": "less-than",
  "<=": "less-or-equal",
  ">": "greater-than",
  ">=": "greater-or-equal",
  "= *": "exists",
  // first match
  "=*": "exists",
  "!= *": "not-exists",
  // first match
  "!=*": "not-exists",
  "~": "matches-phrase",
  "!~": "not-matches-phrase",
  "*~": "search",
  "* ~": "search",
  in: "in",
  "not in": "not in",
  contains: "contains",
  "starts-with": "starts-with",
  "ends-with": "ends-with",
  "not-contains": "not-contains",
  "not-starts-with": "not-starts-with",
  "not-ends-with": "not-ends-with"
};
const getOperatorName = (operator) => OperatorsWithNames[operator] ?? "";
const getOperatorShortName = (comparisonOperator) => {
  for (const [key, value] of Object.entries(OperatorsWithNames)) {
    if (value === comparisonOperator) {
      return key;
    }
  }
  return void 0;
};
const baseOperators = [
  "equals",
  "not-equals",
  "exists",
  "not-exists"
];
const stringOperators = [
  "matches-phrase",
  "not-matches-phrase",
  "contains",
  "starts-with",
  "ends-with",
  "not-contains",
  "not-starts-with",
  "not-ends-with"
];
const listOperators = ["in", "not in"];
const numericOperators = [
  "less-than",
  "less-or-equal",
  "greater-than",
  "greater-or-equal"
];
const getAllowedOperatorsForType = (type) => {
  switch (type) {
    case "String":
      return [...baseOperators, ...listOperators, ...stringOperators];
    case "Number":
      return [...baseOperators, ...numericOperators, ...listOperators];
    case "Boolean":
      return [...baseOperators];
    case "Timestamp":
    case "IPAddress":
    case "Duration":
      return [...baseOperators, ...numericOperators, ...listOperators];
    case "UID":
    case "SmartscapeId":
      return [...baseOperators, ...listOperators];
    case "JSONPath":
    case "Any":
    default:
      return [
        ...baseOperators,
        ...numericOperators,
        ...listOperators,
        ...stringOperators
      ];
  }
};
const FILTERFIELD_DEFAULT_OPERATORS = [
  "equals",
  "not-equals",
  "greater-than",
  "greater-or-equal",
  "less-than",
  "less-or-equal",
  "in",
  "not in",
  "starts-with",
  "contains",
  "ends-with",
  "not-starts-with",
  "not-contains",
  "not-ends-with",
  "exists",
  "not-exists"
];
export {
  FILTERFIELD_DEFAULT_OPERATORS,
  getAllowedOperatorsForType,
  getOperatorName,
  getOperatorShortName
};
