// @dynatrace/strato-components/esm/filters/filter-field/utils/validation-utils/type-guards.js
function isFilterFieldValidatorMapTypeSpecification(valueSpecification) {
  return typeof valueSpecification === "object" && valueSpecification !== null && "type" in valueSpecification && !("value" in valueSpecification);
}
function isFilterFieldValidatorMapDurationSpecification(valueSpecification) {
  const castedValueSpecification = valueSpecification;
  return typeof valueSpecification === "object" && castedValueSpecification.value !== void 0 && castedValueSpecification.unit !== void 0;
}
function isFilterFieldRichValuePredicate(valueSpecification) {
  const castedValueSpecification = valueSpecification;
  return typeof valueSpecification === "object" && castedValueSpecification !== null && "value" in castedValueSpecification && !isFilterFieldValidatorMapTypeSpecification(valueSpecification) && ("displayValue" in castedValueSpecification || "details" in castedValueSpecification);
}
function isFilterFieldKeySuggestionConfig(value) {
  return typeof value === "object" && value !== null && "key" in value && typeof value.key === "string";
}
function isFilterFieldKeySuggestionGroupConfig(value) {
  return typeof value === "object" && value !== null && "suggestions" in value && Array.isArray(value.suggestions) && !("key" in value);
}
function isFilterFieldValueSuggestionGroupConfig(value) {
  return typeof value === "object" && value !== null && "suggestions" in value && Array.isArray(value.suggestions);
}
export {
  isFilterFieldKeySuggestionConfig,
  isFilterFieldKeySuggestionGroupConfig,
  isFilterFieldRichValuePredicate,
  isFilterFieldValidatorMapDurationSpecification,
  isFilterFieldValidatorMapTypeSpecification,
  isFilterFieldValueSuggestionGroupConfig
};
