// @dynatrace/strato-components/esm/filters/filter-field/utils/validation-utils/validation-helper-functions.js
import Colors from "@dynatrace/strato-design-tokens/colors";
import {
  FILTERFIELD_DEFAULT_OPERATORS,
  getAllowedOperatorsForType,
  getOperatorName
} from "./operator-specification.js";
import {
  isFilterFieldRichValuePredicate,
  isFilterFieldValidatorMapDurationSpecification,
  isFilterFieldValidatorMapTypeSpecification,
  isFilterFieldValueSuggestionGroupConfig
} from "./type-guards.js";
import {
  isValidIPAddress,
  isValidSmartscapeId,
  isValidTimestamp,
  isValidUID
} from "./type-validators.js";
import { isDev } from "../../../../core/utils/is-node-environment.js";
import { getValidValueTypes } from "../suggestions-utils/suggestion-type.js";
function isWildCardOperator(currentStatement) {
  const wildCardOperators = [
    "starts-with",
    "ends-with",
    "contains",
    "not-starts-with",
    "not-ends-with",
    "not-contains"
  ];
  return typeof currentStatement?.comparisonOperator?.value === "string" && wildCardOperators.includes(currentStatement?.comparisonOperator?.value);
}
function isValidIPAddressType(valueType, textValue, validTypes, parserConfig) {
  if (!parserConfig?.ipAddressConversion && validTypes.includes("IPAddress")) {
    if (isDev()) {
      console.warn(
        "IPAddress is configured as a type in the validatorMap of the FilterField. To get full support for IPAddress types please also enable IPAddress conversion in the parserConfig."
      );
    }
    return valueType === "Identifier" && isValidIPAddress(textValue);
  } else if (validTypes.includes("IPAddress")) {
    return valueType === "IPAddress";
  }
  return false;
}
function isValidTimestampType(valueType, textValue, validTypes, parserConfig) {
  if (!parserConfig?.timestampConversion && validTypes.includes("Timestamp")) {
    if (isDev()) {
      console.warn(
        "Timestamp is configured as a type in the validatorMap of the FilterField. To get full support for Timestamp types please also enable Timestamp conversion in the parserConfig."
      );
    }
    return (valueType === "Identifier" || valueType === "String") && isValidTimestamp(textValue);
  } else if (validTypes.includes("Timestamp")) {
    return valueType === "Timestamp";
  }
  return false;
}
function isValidUIDType(valueType, textValue, validTypes, parserConfig) {
  if (!parserConfig?.uidConversion && validTypes.includes("UID")) {
    if (isDev()) {
      console.warn(
        "UID is configured as a type in the validatorMap of the FilterField. To get full support for UID types please also enable UID conversion in the parserConfig."
      );
    }
    return valueType === "Identifier" && isValidUID(textValue);
  } else if (validTypes.includes("UID")) {
    return valueType === "UID";
  }
  return false;
}
function isValidSmartscapeIdType(valueType, textValue, validTypes, parserConfig) {
  if (
    // eslint-disable-next-line @typescript-eslint/no-deprecated
    !parserConfig?.smartscapeIdConversion && validTypes.includes("SmartscapeId")
  ) {
    if (isDev()) {
      console.warn(
        "SmartscapeId is configured as a type in the validatorMap of the FilterField. To get full support for SmartscapeId types please also enable SmartscapeId conversion in the parserConfig."
      );
    }
    return valueType === "Identifier" && isValidSmartscapeId(textValue);
  } else if (validTypes.includes("SmartscapeId")) {
    return valueType === "SmartscapeId";
  }
  return false;
}
function checkIsValidType(valueType, valueTextValue, validTypes, parserConfig, customTypes) {
  return validTypes.includes("Any") || validTypes.includes(valueType) || validTypes.includes("String") && valueType === "Identifier" || validTypes.includes("Number") && valueType === "BigInt" || isValidIPAddressType(valueType, valueTextValue, validTypes, parserConfig) || isValidUIDType(valueType, valueTextValue, validTypes, parserConfig) || isValidSmartscapeIdType(
    valueType,
    valueTextValue,
    validTypes,
    parserConfig
  ) || isValidTimestampType(valueType, valueTextValue, validTypes, parserConfig) || !!isValidCustomType(valueTextValue, customTypes);
}
function hasValuePredicate(entry) {
  return !!entry && typeof entry === "object" && typeof entry.valuePredicate === "function";
}
function isValidCustomType(textValue, customTypes) {
  if (!customTypes) {
    return false;
  }
  for (const [typeName, validatorFn] of Object.entries(customTypes)) {
    if (typeof validatorFn === "function") {
      if (validatorFn(textValue)) {
        return typeName;
      }
      continue;
    }
    if (hasValuePredicate(validatorFn) && validatorFn.valuePredicate(textValue)) {
      return typeName;
    }
  }
  return false;
}
function isValidOperator(textValue, key, internalValidatorMap, parserConfig) {
  const operatorName = getOperatorName(textValue);
  if (operatorName === "") {
    return false;
  }
  if (operatorName === "search" && !parserConfig?.searchConversion) {
    return true;
  }
  return getValidOperators(key, internalValidatorMap, parserConfig).includes(
    operatorName
  );
}
function isValidValue(key, valueTextValue, valueType, internalValidatorMap, currentStatement, filterFieldValue, customTypes, parserConfig) {
  if (!key || !internalValidatorMap) {
    return { isValid: true };
  }
  const predicate = internalValidatorMap.normalizedKeys.byKey.get(key);
  if (!predicate) {
    return { isValid: true };
  }
  const validValues = predicate.valuePredicate;
  if (typeof validValues === "function") {
    const result = validValues({
      currentStatement,
      filterFieldValue
    });
    if (typeof result === "boolean") {
      return { isValid: result };
    }
    return { isValid: false, message: result };
  }
  if (valueType === "Variable") {
    return { isValid: true };
  }
  let validTypes;
  let hasLiteralValues;
  if (predicate.valueType) {
    validTypes = Array.isArray(predicate.valueType) ? predicate.valueType : [predicate.valueType];
    hasLiteralValues = Array.isArray(validValues);
  } else if (validValues !== void 0) {
    ({ types: validTypes, hasLiteralValues } = getValidValueTypes(validValues));
  } else {
    return { isValid: true };
  }
  const isValidType = checkIsValidType(
    valueType,
    valueTextValue,
    validTypes,
    parserConfig,
    customTypes
  );
  if (Array.isArray(validValues)) {
    let textMatches = false;
    for (const validValue of validValues) {
      if (isFilterFieldValidatorMapTypeSpecification(validValue)) {
        if (!isValidType && !hasLiteralValues || isValidType) {
          return { isValid: isValidType, validTypes };
        }
      } else if (isFilterFieldValueSuggestionGroupConfig(validValue)) {
        if (validValue.suggestions.some((suggestion) => {
          if (isFilterFieldValidatorMapTypeSpecification(suggestion)) {
            return false;
          }
          return literalMatchesTextValue(valueTextValue, suggestion);
        })) {
          textMatches = true;
        }
        continue;
      } else if (literalMatchesTextValue(valueTextValue, validValue)) {
        textMatches = true;
      }
    }
    return {
      isValid: textMatches || isWildCardOperator(currentStatement),
      validTypes
    };
  }
  if (isFilterFieldValidatorMapTypeSpecification(validValues)) {
    return { isValid: isValidType, validTypes };
  }
  if (isFilterFieldValueSuggestionGroupConfig(validValues)) {
    const groupSuggestionMatches = validValues.suggestions.some(
      (suggestion) => {
        if (isFilterFieldValidatorMapTypeSpecification(suggestion)) {
          return false;
        }
        return literalMatchesTextValue(valueTextValue, suggestion);
      }
    );
    return {
      isValid: groupSuggestionMatches || isWildCardOperator(currentStatement)
    };
  }
  if (validValues === void 0) {
    return { isValid: isValidType };
  }
  return {
    isValid: literalMatchesTextValue(valueTextValue, validValues),
    validTypes
  };
}
function literalMatchesTextValue(textValue, predicate) {
  if (isFilterFieldValidatorMapDurationSpecification(predicate)) {
    return textValue === `${predicate.value}${predicate.unit}`;
  }
  if (isFilterFieldRichValuePredicate(predicate)) {
    const cleaned2 = `${predicate.value}`.replaceAll(/(?<!\\)(?:\\\\)*(\\)(?!\\)/g, "").replaceAll(/(\\\\)/g, "\\");
    return textValue === cleaned2;
  }
  const cleaned = `${predicate}`.replaceAll(/(?<!\\)(?:\\\\)*(\\)(?!\\)/g, "").replaceAll(/(\\\\)/g, "\\");
  return textValue === cleaned;
}
function getValuePredicateTextValue(predicate) {
  if (isFilterFieldValidatorMapDurationSpecification(predicate)) {
    return `${predicate.value}${predicate.unit}`;
  } else if (isFilterFieldRichValuePredicate(predicate)) {
    return `${predicate.value}`;
  } else {
    return `${predicate}`;
  }
}
function getValidOperators(key, internalValidatorMap, parserConfig) {
  key = typeof key === "string" ? key.replace(/^"(.*?(?<!\\)(?:\\\\)*)"$/, "$1") : key;
  const defaultOperators = [
    ...FILTERFIELD_DEFAULT_OPERATORS,
    // eslint-disable-next-line @typescript-eslint/no-deprecated
    ...parserConfig?.searchConversion ? ["search"] : []
  ];
  if (!internalValidatorMap) {
    return defaultOperators;
  }
  const globalOperators = internalValidatorMap.comparisonOperators || defaultOperators;
  const predicate = key ? internalValidatorMap.normalizedKeys.byKey.get(key) : void 0;
  if (!predicate && key === "*") {
    return [];
  }
  if (!predicate) {
    return globalOperators;
  }
  if (predicate.comparisonOperators) {
    return predicate.comparisonOperators;
  }
  let allowedTypes;
  if (predicate.valueType) {
    allowedTypes = Array.isArray(predicate.valueType) ? predicate.valueType : [predicate.valueType];
  } else if (!predicate.valuePredicate || typeof predicate.valuePredicate === "function") {
    allowedTypes = "Any";
  } else {
    allowedTypes = getValidValueTypes(predicate.valuePredicate).types;
  }
  if (allowedTypes !== "Any") {
    return intersectOperators(
      globalOperators,
      allowedTypes.reduce((acc, type) => {
        acc.push(...getAllowedOperatorsForType(type));
        return acc;
      }, [])
    );
  }
  return globalOperators;
}
const OPERATOR_SHORT_NAME_MAP = {
  equals: "=",
  "not-equals": "!=",
  "less-than": "<",
  "less-or-equal": "<=",
  "greater-than": ">",
  "greater-or-equal": ">=",
  "matches-phrase": "~",
  "not-matches-phrase": "!~",
  in: "in",
  "not in": "not in",
  contains: "contains",
  "starts-with": "starts-with",
  "ends-with": "ends-with",
  "not-starts-with": "not-starts-with",
  "not-contains": "not-contains",
  "not-ends-with": "not-ends-with",
  exists: "= *",
  "not-exists": "!= *"
};
function getValidOperatorsShort(key, internalValidatorMap, parserConfig) {
  const operators = getValidOperators(key, internalValidatorMap, parserConfig);
  const result = [];
  for (const [key2, value] of Object.entries(OPERATOR_SHORT_NAME_MAP)) {
    if (operators.includes(key2)) {
      result.push(value);
    }
  }
  return result;
}
function getValidOperatorsColored(key, internalValidatorMap, parserConfig) {
  return getValidOperatorsShort(key, internalValidatorMap, parserConfig).map(
    (operator) => `<span style="color: ${Colors.Syntax.Operators}">${operator}</span>`
  ).join(", ");
}
function getColoredSearchOperator() {
  return `<span style="color: ${Colors.Syntax.Variable}">*</span> <span style="color: ${Colors.Syntax.Operators}">~</span>`;
}
function intersectOperators(globalOperators, operators) {
  return globalOperators.filter((op) => operators.includes(op));
}
export {
  getColoredSearchOperator,
  getValidOperators,
  getValidOperatorsColored,
  getValidOperatorsShort,
  getValuePredicateTextValue,
  isValidCustomType,
  isValidOperator,
  isValidValue,
  isWildCardOperator
};
