// @dynatrace/strato-components/charts/meter-bar/multi-meter-bar/default.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var default_exports = {};
__export(default_exports, {
  DEFAULT_COLOR_PALETTE: () => DEFAULT_COLOR_PALETTE,
  defaultConfig: () => defaultConfig
});
module.exports = __toCommonJS(default_exports);
var import_constants = require("../shared/constants.js");
const DEFAULT_COLOR_PALETTE = "categorical";
const TOOLTIP_VARIANT_DEFAULT = "single";
const defaultConfig = {
  colorPalette: DEFAULT_COLOR_PALETTE,
  size: import_constants.SIZE_DEFAULT,
  min: import_constants.MIN_DEFAULT_VALUE,
  max: import_constants.MAX_DEFAULT_VALUE,
  tooltip: { hidden: false, variant: TOOLTIP_VARIANT_DEFAULT },
  formatter: import_constants.DEFAULT_FORMATTER,
  legend: false,
  loading: false
};
