// @dynatrace/strato-components/charts/meter-bar/multi-meter-bar/MultiMeterBarChart.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MultiMeterBarChart_exports = {};
__export(MultiMeterBarChart_exports, {
  MultiMeterBarChart: () => MultiMeterBarChart
});
module.exports = __toCommonJS(MultiMeterBarChart_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_MultiMeterBarChartLegend = require("./components/MultiMeterBarChartLegend.js");
var import_MultiMeterBarChartPlotArea = require("./components/MultiMeterBarChartPlotArea.js");
var import_constants = require("./constants.js");
var import_useMultiMeterBarChartConfig = require("./hooks/useMultiMeterBarChartConfig.js");
var import_MultiMeterBarChartProviders = require("./providers/MultiMeterBarChartProviders.js");
var import_Label = require("./slots/Label.js");
var import_Legend = require("./slots/Legend.js");
var import_Max = require("./slots/Max.js");
var import_Min = require("./slots/Min.js");
var import_Segment = require("./slots/Segment.js");
var import_Tooltip = require("./slots/Tooltip.js");
var import_Value = require("./slots/Value.js");
var import_useBehavioralTrackingProps = require("../../../core/hooks/useBehavioralTrackingProps.js");
var import_Intent = require("../../../core/slots/Intent/Intent.js");
var import_defaultChartValues_css = require("../../../core/styles/defaultChartValues.sty.js");
var import_ColorRule = require("../../core/slots/ColorRule.js");
var import_chart_types = require("../../core/types/chart-types.js");
var import_MeterBarChartFormattedValue = require("../shared/components/formatted-value/MeterBarChartFormattedValue.js");
var import_MeterBarLayout = require("../shared/components/meter-bar-layout/MeterBarLayout.js");
var import_constants2 = require("../shared/constants.js");
var import_Toolbar = require("../shared/slots/Toolbar.js");
const COMPONENT_NAME = "MultiMeterBarChart";
const MultiMeterBarChartRoot = (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    width,
    height,
    className,
    style,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "data-testid": dataTestId = "multi-meter-bar-chart-root",
    seriesActions,
    prefixIcon
  } = props;
  const [behavioralTrackingProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(props);
  const config = (0, import_useMultiMeterBarChartConfig.useMultiMeterBarChartConfig)(props);
  const {
    colorPalette,
    size = import_constants2.SIZE_DEFAULT,
    min = import_constants2.MIN_DEFAULT_VALUE,
    max,
    loading,
    formatter,
    unit,
    labelSlot,
    valueSlot,
    minSlot,
    maxSlot,
    segments,
    legend,
    tooltip,
    totalValue,
    colorRules,
    toolbar
  } = config;
  const multiMeterBarChartRef = (0, import_react.useRef)(null);
  const consumerConfig = {
    type: import_chart_types.ChartType.MultiMeterBar,
    colorPalette,
    size,
    min,
    max,
    formatter,
    unit,
    tooltip: { enabled: !tooltip.hidden, variant: tooltip.variant },
    legend,
    loading
  };
  const label = labelSlot ?? "";
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_MultiMeterBarChartProviders.MultiMeterBarChartProviders,
    {
      colorPalette,
      formatter,
      unit,
      segments,
      colorRules,
      toolbar,
      loading,
      seriesActions,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
        import_MeterBarLayout.MeterBarLayout,
        {
          ...behavioralTrackingProps,
          width,
          height,
          ref: multiMeterBarChartRef,
          size,
          className: (0, import_clsx.default)(import_defaultChartValues_css.defaultChartSize, className),
          style,
          "data-dtrum-mask": dataDtrumMask,
          "data-dtrum-allow": dataDtrumAllow,
          "data-testid": dataTestId,
          "aria-label": import_constants.MULTIMETER_BAR_CHART_ARIA_LABEL,
          "data-dt-component": COMPONENT_NAME,
          consumerConfig,
          userRef: forwardedRef,
          children: [
            prefixIcon && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_MeterBarLayout.MeterBarLayout.Icon, { children: prefixIcon }),
            labelSlot && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_MeterBarLayout.MeterBarLayout.Label, { children: labelSlot }),
            valueSlot && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_MeterBarLayout.MeterBarLayout.Value, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_MeterBarChartFormattedValue.MeterBarChartFormattedValue,
              {
                value: totalValue,
                total: max - min,
                auto: valueSlot?.auto,
                children: valueSlot?.value
              }
            ) }),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_MeterBarLayout.MeterBarLayout.Chart, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_MultiMeterBarChartPlotArea.MultiMeterBarChartPlotArea,
              {
                label,
                max,
                min,
                tooltipConfig: tooltip
              }
            ) }),
            minSlot && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_MeterBarLayout.MeterBarLayout.Min, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_MeterBarChartFormattedValue.MeterBarChartFormattedValue, { value: min, auto: minSlot?.auto, children: minSlot?.value }) }),
            maxSlot && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_MeterBarLayout.MeterBarLayout.Max, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_MeterBarChartFormattedValue.MeterBarChartFormattedValue, { value: max, auto: maxSlot?.auto, children: maxSlot?.value }) }),
            legend && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_MeterBarLayout.MeterBarLayout.Legend, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_MultiMeterBarChartLegend.MultiMeterBarChartLegend, {}) })
          ]
        }
      )
    }
  );
});
MultiMeterBarChartRoot.displayName = "_MultiMeterBarChart";
const MultiMeterBarChart = Object.assign(MultiMeterBarChartRoot, {
  Label: import_Label.Label,
  Legend: import_Legend.Legend,
  Max: import_Max.Max,
  Min: import_Min.Min,
  Segment: import_Segment.Segment,
  Value: import_Value.Value,
  Tooltip: import_Tooltip.Tooltip,
  ColorRule: import_ColorRule.ColorRule,
  Intent: import_Intent.Intent,
  Toolbar: import_Toolbar.Toolbar
});
