// @dynatrace/strato-components/esm/charts/core/components/toolbar/ChartToolbarRenderer.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import {
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState
} from "react";
import { useIntl } from "react-intl";
import {
  ChevronRightDoubleIcon,
  DragAllDirectionIcon
} from "@dynatrace/strato-icons";
import {
  chartToolbar,
  chartToolbarArea,
  defaultOverlayPadding,
  grip,
  outerContainer
} from "./ChartToolbarRenderer.sty.js";
import { ChartToolbarVisibilityOverlay } from "./ChartToolbarVisibilityOverlay.js";
import { CollapseButton } from "./CollapseButton.js";
import { ChartDataMenu } from "./components/ChartDataMenu.js";
import { Control } from "./components/Control.js";
import { HorizontalDivider, VerticalDivider } from "./components/divider.js";
import { IntentsCollapsedAndDownloadSection } from "./components/IntentsCollapsedAndDownloadSection.js";
import { ToolbarMenu } from "./components/ToolbarMenu.js";
import { TOOLBAR_DRAG_BUTTON_ARIA_LABEL } from "./constants.js";
import {
  ChartToolbarContext
} from "./context/ChartToolbar.context.js";
import { useToolbar } from "./hooks/useToolbar.js";
import { TOOLBAR_MESSAGES } from "./messages.js";
import { doesFitInBounds } from "./utils/does-fit-in-bounds.js";
import { estimateToolbarWidth } from "./utils/estimate-toolbar-width.js";
import { getToolbarPlacement } from "./utils/get-toolbar-placement.js";
import { Button } from "../../../../buttons/button/Button.js";
import { useBehavioralTrackingProps as _useBehavioralTrackingProps } from "../../../../core/hooks/useBehavioralTrackingProps.js";
import { useMergeRefs } from "../../../../core/hooks/useMergeRefs.js";
import { Flex } from "../../../../layouts/flex/Flex.js";
import { Tooltip } from "../../../../overlays/tooltip/Tooltip.js";
import { A11yToolbarTabIndexContext } from "../../context/A11yToolbarTabIndexContext.js";
import { useA11yToolbarControls } from "../../hooks/useA11yToolbarControls.js";
import { useDragElement } from "../../hooks/useDragElement.js";
import { useObserveContainerSize } from "../../hooks/useObserveContainerSize.js";
import { getDraggableBounds } from "../../utils/get-draggable-bounds.js";
import { INTERACTION_MODE } from "../interactions-overlay/types.js";
const InternalChartToolbarRenderer = (props) => {
  const [
    behaviorTrackingProps,
    {
      children,
      visible,
      placement,
      collapsed = false,
      draggable = true,
      removeDragHandle = false,
      initialA11yState,
      zIndex,
      onToggleToolbar,
      onToggleMenu,
      onMouseEnter,
      onMouseLeave,
      menuIcon,
      menuTriggerLabel,
      menuDisplayMode = "collapsed",
      onClose,
      isDownloadEnabled = false,
      intents,
      handleDownloadDataToolbar,
      toolbarDisplayMode = "overlay",
      visibleWhenMouseInBounds = false,
      overlayClassName = defaultOverlayPadding,
      intentsAsyncTemplate,
      handlerRef
    }
  ] = _useBehavioralTrackingProps(props);
  const {
    isExpanded,
    lastItemSelected,
    updateLastItemSelected,
    toggleExpanded,
    isCollapsable,
    fitsBounds,
    updateFitsBounds,
    shouldShowMenu
  } = useToolbar(collapsed, menuDisplayMode);
  const isInlineMode = toolbarDisplayMode === "inline";
  const [isMenuOpen, setIsMenuOpen] = useState(void 0);
  const [toolbarElement, setToolbarElement] = useState();
  const [boundSize, boundRef] = useObserveContainerSize();
  const [toolbarSize, toolbarRefCallback] = useObserveContainerSize();
  const intl = useIntl();
  const mouseInBounds = useRef(false);
  const [isVisible, setIsVisible] = useState(false);
  const toolbarElementRef = useCallback((toolbarElement2) => {
    setToolbarElement(toolbarElement2);
  }, []);
  const toolbarRef = useMergeRefs([toolbarRefCallback, toolbarElementRef]);
  const { a11yControlsValue, setKeyEvent, setActiveControl } = useA11yToolbarControls(initialA11yState);
  const maxBounds = useMemo(
    () => getDraggableBounds(toolbarSize, boundSize),
    [toolbarSize, boundSize]
  );
  useEffect(() => {
    updateFitsBounds(
      doesFitInBounds(estimateToolbarWidth(children), boundSize)
    );
  }, [children, boundSize]);
  useEffect(() => {
    toggleExpanded(fitsBounds && !collapsed);
  }, [collapsed, fitsBounds]);
  const initialPosition = getToolbarPlacement(placement, maxBounds);
  const { isDragging, triggerDragRef, didDrag, setDidDrag } = useDragElement(
    toolbarElement,
    maxBounds,
    initialPosition
  );
  const expandHandler = useCallback(
    (isExpanded2) => {
      toggleExpanded(isExpanded2);
      onToggleToolbar?.(isExpanded2);
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [onToggleToolbar]
  );
  const handleToggleMenu = (open) => {
    if (!open) {
      if (!mouseInBounds.current) {
        onMouseLeave?.();
      } else {
        onMouseEnter?.();
      }
    }
    setIsMenuOpen(open);
    onToggleMenu?.(open);
  };
  const collapseButton = useMemo(
    () => Array.isArray(children) && /* @__PURE__ */ jsx(
      CollapseButton,
      {
        onClick: (isExpanded2) => {
          expandHandler(isExpanded2);
          setActiveControl("explore");
        }
      }
    ),
    [children, expandHandler, setActiveControl]
  );
  const expandButton = useMemo(
    () => /* @__PURE__ */ jsx(
      Control,
      {
        prefixIcon: /* @__PURE__ */ jsx(ChevronRightDoubleIcon, {}),
        tooltipText: intl.formatMessage(TOOLBAR_MESSAGES.chartToolbarExpand),
        onClick: () => expandHandler(true),
        tabIndex: -1,
        testId: "toolbar-expand-button",
        "data-dt-name": "ToolbarExpand"
      }
    ),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [expandHandler]
  );
  const onMouseEnterHandler = () => {
    mouseInBounds.current = true;
    onMouseEnter?.();
  };
  const onMouseLeaveHandler = (event) => {
    event.stopPropagation();
    mouseInBounds.current = false;
    onMouseLeave?.();
  };
  const handleOnMenuEnter = () => {
    mouseInBounds.current = false;
  };
  const handleOnMenuLeave = () => {
    mouseInBounds.current = true;
  };
  const onCloseAutoFocus = (e) => {
    if ((lastItemSelected === INTERACTION_MODE.Explore && onClose || lastItemSelected === INTERACTION_MODE.Zoom) && onClose) {
      e.preventDefault();
      onClose();
      updateLastItemSelected(void 0);
    }
  };
  const handleResetDidDrag = useCallback(() => {
    setDidDrag(false);
  }, [setDidDrag]);
  const showCollapsedIntentAndDownloadSubmenu = (!isExpanded || !fitsBounds) && (!!intents?.length || isDownloadEnabled);
  const showExpandButton = !isExpanded && fitsBounds;
  const showCollapseButton = isExpanded && isCollapsable;
  const showIntentMenu = isExpanded && fitsBounds && !!(intents?.length || intentsAsyncTemplate);
  const toolbarMenuConfig = useMemo(
    () => ({
      isMenuElement: true,
      isExpanded,
      lastItemSelected,
      updateLastItemSelected
    }),
    [isExpanded, lastItemSelected, updateLastItemSelected]
  );
  const intentsMenuConfig = useMemo(
    () => ({
      isMenuElement: false,
      isExpanded,
      lastItemSelected,
      didDrag,
      resetDidDrag: handleResetDidDrag,
      updateLastItemSelected
    }),
    [
      isExpanded,
      lastItemSelected,
      didDrag,
      handleResetDidDrag,
      updateLastItemSelected
    ]
  );
  const handleKeyDown = (event) => {
    if (event.target === triggerDragRef.current && event.code !== "Tab") {
      event.stopPropagation();
      event.preventDefault();
      return;
    }
    if (event.code === "ArrowLeft" || event.code === "ArrowRight") {
      event.stopPropagation();
      setKeyEvent(event);
    }
    if (event.code === "Enter") {
      event.stopPropagation();
    }
  };
  const handleOnBlur = (event) => {
    event.stopPropagation();
  };
  return /* @__PURE__ */ jsx(
    ChartToolbarVisibilityOverlay,
    {
      handlerRef,
      boundSize,
      setIsVisible,
      visibleWhenMouseInBounds,
      isVisible,
      children: /* @__PURE__ */ jsx(
        "div",
        {
          className: clsx(
            chartToolbarArea({
              inline: isInlineMode
            }),
            overlayClassName
          ),
          "data-testid": "chart-plot-overlay-toolbar",
          ref: boundRef,
          style: { zIndex },
          children: /* @__PURE__ */ jsx(
            "div",
            {
              className: chartToolbar({ inline: isInlineMode }),
              id: "chart-toolbar",
              ref: toolbarRef,
              role: "toolbar",
              "data-testid": "chart-toolbar-container",
              onBlur: handleOnBlur,
              onKeyDown: handleKeyDown,
              onMouseEnter: onMouseEnterHandler,
              onMouseLeave: onMouseLeaveHandler,
              ...behaviorTrackingProps,
              children: /* @__PURE__ */ jsx(A11yToolbarTabIndexContext.Provider, { value: a11yControlsValue, children: /* @__PURE__ */ jsx(
                "div",
                {
                  className: outerContainer({
                    visible: visible || isMenuOpen || isDragging || visibleWhenMouseInBounds && isVisible,
                    inline: isInlineMode
                  }),
                  "data-testid": "chart-toolbar-wrapper",
                  children: /* @__PURE__ */ jsxs(Flex, { gap: 0, flexDirection: "row", alignItems: "center", children: [
                    draggable && !removeDragHandle && [
                      /* @__PURE__ */ jsx(
                        Tooltip,
                        {
                          text: intl.formatMessage(
                            TOOLBAR_MESSAGES.chartToolbarMove
                          ),
                          disabled: isDragging,
                          children: /* @__PURE__ */ jsx(
                            Button,
                            {
                              className: grip,
                              ref: triggerDragRef,
                              variant: "default",
                              size: "condensed",
                              "data-testid": "chart-toolbar-grip",
                              "aria-label": TOOLBAR_DRAG_BUTTON_ARIA_LABEL,
                              children: /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(DragAllDirectionIcon, {}) })
                            }
                          )
                        },
                        "tooltip"
                      ),
                      /* @__PURE__ */ jsx(VerticalDivider, {}, "divider")
                    ],
                    /* @__PURE__ */ jsxs(
                      "div",
                      {
                        "data-testid": "toolbar-children",
                        style: { display: "flex", alignItems: "center" },
                        ref: draggable && removeDragHandle ? triggerDragRef : void 0,
                        children: [
                          shouldShowMenu && /* @__PURE__ */ jsx(ChartToolbarContext.Provider, { value: toolbarMenuConfig, children: /* @__PURE__ */ jsxs(
                            ToolbarMenu,
                            {
                              onMenuOpen: handleToggleMenu,
                              activeModeIcon: menuIcon,
                              triggerLabel: menuTriggerLabel,
                              onMenuEnter: handleOnMenuEnter,
                              onMenuLeave: handleOnMenuLeave,
                              menuAlignment: isExpanded ? "start" : "end",
                              onCloseAutoFocus,
                              "data-dt-name": "ToolbarTools",
                              "data-dt-properties": isExpanded && fitsBounds ? "state:changeMode" : "state:tools",
                              children: [
                                children,
                                /* @__PURE__ */ jsx(
                                  IntentsCollapsedAndDownloadSection,
                                  {
                                    hidden: !showCollapsedIntentAndDownloadSubmenu,
                                    intents,
                                    showDownloadButton: isDownloadEnabled,
                                    handleDownloadDataToolbar
                                  }
                                ),
                                showExpandButton && /* @__PURE__ */ jsxs(Fragment, { children: [
                                  /* @__PURE__ */ jsx(HorizontalDivider, {}),
                                  expandButton
                                ] })
                              ]
                            }
                          ) }),
                          /* @__PURE__ */ jsxs(ChartToolbarContext.Provider, { value: intentsMenuConfig, children: [
                            children,
                            showIntentMenu && /* @__PURE__ */ jsx(
                              ChartDataMenu,
                              {
                                onMenuOpen: handleToggleMenu,
                                onMenuEnter: handleOnMenuEnter,
                                onMenuLeave: handleOnMenuLeave,
                                isDownloadEnabled,
                                handleDownloadDataToolbar,
                                intents,
                                showDivider: !!menuIcon,
                                isDragging,
                                buttonMode: isInlineMode ? "inline" : "standard",
                                intentsAsyncTemplate
                              }
                            )
                          ] }),
                          showCollapseButton && collapseButton
                        ]
                      }
                    )
                  ] })
                }
              ) })
            }
          )
        }
      )
    }
  );
};
InternalChartToolbarRenderer["displayName"] = "ChartToolbarRenderer";
const ChartToolbarRenderer = Object.assign(
  InternalChartToolbarRenderer,
  {
    Control
  }
);
export {
  ChartToolbarRenderer,
  InternalChartToolbarRenderer
};
