// @dynatrace/strato-components/esm/charts/core/components/toolbar/ChartToolbarVisibilityOverlay.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import { useEffect, useMemo, useRef } from "react";
import { toolbarVisibilityOverlay } from "./ChartToolbarRenderer.sty.js";
import { useApplyCallbackOnScroll } from "../../hooks/useApplyCallbackOnScroll.js";
import { useEffectOnValueChange } from "../../hooks/useEffectOnValueChange.js";
import { isWithinBounds } from "../../utils/is-within-bounds.js";
import { interactionsOverlay } from "../interactions-overlay/InteractionsOverlay.sty.js";
const ChartToolbarVisibilityOverlay = (props) => {
  const { children, visibleWhenMouseInBounds } = props;
  if (!visibleWhenMouseInBounds) {
    return children;
  }
  return /* @__PURE__ */ jsx(VisibilityOverlay, { ...props });
};
const VisibilityOverlay = (props) => {
  const { boundSize, children, setIsVisible, isVisible, handlerRef } = props;
  const overlayRef = useRef(null);
  const overlayState = useMemo(() => {
    let bounds;
    return {
      inBounds: (e) => {
        if (!bounds) {
          bounds = overlayRef.current?.getBoundingClientRect();
        }
        return isWithinBounds(e, bounds);
      },
      resetBounds: () => {
        bounds = void 0;
      }
    };
  }, []);
  useApplyCallbackOnScroll(!isVisible, () => {
    overlayState.resetBounds();
  });
  useEffectOnValueChange(() => {
    overlayState.resetBounds();
    setIsVisible(false);
  }, [boundSize.height, boundSize.width]);
  useEffect(() => {
    if (handlerRef?.current) {
      handlerRef.current.handleFocus = () => {
        setIsVisible(true);
      };
    }
    return () => {
      if (handlerRef?.current) {
        handlerRef.current.handleFocus = () => null;
      }
    };
  });
  useEffect(() => {
    const cbMouse = (event) => {
      if (!overlayState.inBounds(event)) {
        setIsVisible(false);
        overlayState.resetBounds();
      }
    };
    if (isVisible) {
      document.addEventListener("mouseup", cbMouse);
      document.addEventListener("pointerout", cbMouse);
    }
    return () => {
      document.removeEventListener("mouseup", cbMouse);
      document.removeEventListener("pointerout", cbMouse);
    };
  }, [isVisible]);
  return /* @__PURE__ */ jsx(
    "div",
    {
      role: "none",
      ref: overlayRef,
      className: clsx(toolbarVisibilityOverlay, interactionsOverlay),
      onPointerMove: (e) => {
        if (overlayState.inBounds(e)) {
          isVisible || setIsVisible(true);
        }
      },
      onClick: () => {
        isVisible || setIsVisible(true);
      },
      children
    }
  );
};
export {
  ChartToolbarVisibilityOverlay
};
