// @dynatrace/strato-components/esm/charts/core/components/toolbar/CollapseButton.js
import { jsx } from "react/jsx-runtime";
import { useRef } from "react";
import { useIntl } from "react-intl";
import { ChevronLeftDoubleIcon } from "@dynatrace/strato-icons";
import { Control } from "./components/Control.js";
import { TOOLBAR_MESSAGES } from "./messages.js";
import { useA11yToolbarTabIndex } from "../../hooks/useA11yToolbarTabIndex.js";
const CollapseButton = ({ onClick }) => {
  const buttonRef = useRef(null);
  const intl = useIntl();
  const tabIndex = useA11yToolbarTabIndex("collapse");
  const onClickHandler = () => {
    onClick(false);
  };
  const onKeyDownHandler = (event) => {
    if (event.key === "Enter") {
      onClick(false);
    }
  };
  return /* @__PURE__ */ jsx(
    Control,
    {
      ref: buttonRef,
      testId: "toolbar-collapse-button",
      prefixIcon: /* @__PURE__ */ jsx(ChevronLeftDoubleIcon, {}),
      tooltipText: intl.formatMessage(TOOLBAR_MESSAGES.chartToolbarCollapse),
      onKeyDown: onKeyDownHandler,
      onClick: onClickHandler,
      tabIndex,
      divider: "left",
      "data-dt-name": "ToolbarCollapse"
    }
  );
};
CollapseButton["displayName"] = "CollapseButton";
export {
  CollapseButton
};
