// @dynatrace/strato-components/filters/timeframe-selector/components/trigger/Trigger.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Trigger_exports = {};
__export(Trigger_exports, {
  Trigger: () => Trigger,
  isTimeframeSelectorTriggerElement: () => isTimeframeSelectorTriggerElement
});
module.exports = __toCommonJS(Trigger_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_is = require("react-is");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_DisplayValue = require("./DisplayValue.js");
var styles = __toESM(require("./Trigger.sty.js"));
var import_Button = require("../../../../buttons/button/Button.js");
var import_is_visual_refresh_enabled = require("../../../../core/utils/is-visual-refresh-enabled.js");
var import_merge_props = require("../../../../core/utils/merge-props.js");
var import_useTimeframeSelectorContext = require("../../hooks/useTimeframeSelectorContext.js");
function isTimeframeSelectorTriggerElement(child) {
  return (0, import_react_is.isElement)(child) && child.type === Trigger;
}
function getDisplayValueSlot(children) {
  let displayValue = null;
  import_react.Children.forEach(children, (child) => {
    if ((0, import_DisplayValue.isTimeframeSelectorDisplayValueElement)(child)) {
      displayValue = displayValue || [];
      displayValue.push(child);
    }
  });
  return displayValue;
}
const Trigger = (props) => {
  const {
    children,
    placeholder,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    ...remainingProps
  } = props;
  const displayValueSlot = (0, import_react.useMemo)(
    () => getDisplayValueSlot(children),
    [children]
  );
  const {
    triggerProps,
    displayValue: contextDisplayValue,
    details,
    readOnly,
    showError,
    disabled
  } = (0, import_useTimeframeSelectorContext.useTimeframeSelectorContext)();
  const isCritical = (0, import_is_visual_refresh_enabled.isVisualRefreshEnabled)() ? showError && !disabled : showError;
  const color = isCritical ? "critical" : "neutral";
  const displayValue = displayValueSlot ?? contextDisplayValue;
  const placeholderIfEmpty = placeholder ?? contextDisplayValue;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_Button.Button,
    {
      ...(0, import_merge_props.mergeProps)(triggerProps, remainingProps),
      "data-testid": dataTestId,
      variant: "emphasized",
      color,
      textAlign: "start",
      style: consumerStyle,
      className: (0, import_clsx.default)(consumerClassName, {
        [styles.readOnly]: readOnly,
        [styles.readOnlyColor[color]]: readOnly
      }),
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Suffix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_strato_icons.ChevronDownSmallIcon,
          {
            className: (0, import_clsx.default)(
              styles.suffix,
              triggerProps?.["aria-expanded"] && styles.suffixOpen
            )
          }
        ) }),
        details ? displayValue : placeholderIfEmpty
      ]
    }
  );
};
Trigger.displayName = "TimeframeSelector.Trigger";
