// @dynatrace/strato-components/esm/charts/core/components/selection-area/SelectionTooltipWithAsyncActions.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { debounce } from "lodash-es";
import {
  useCallback,
  useEffect,
  useState
} from "react";
import { CustomActions, CustomIntents } from "./tooltip/CustomActions.js";
import { getStratoLogger } from "../../../../core/sdk/sdk-logger.js";
import { Divider } from "../../../../layouts/divider/Divider.js";
import { OverlayTooltip } from "../overlay-tooltip/OverlayTooltip.js";
import { divider } from "../overlay-tooltip/tooltip.sty.js";
const DEBOUNCING_TIME = 150;
const emptyActionsMenu = { actions: [], intents: [] };
const SelectionTooltipWithAsyncActions = (props) => {
  const [actionsMenu, setActionsMenu] = useState(emptyActionsMenu);
  const logger = getStratoLogger("SelectionTooltipWithAsyncActions");
  const fetchActions = useCallback(
    debounce(
      (actions) => {
        const promise = actions();
        if (promise) {
          promise.then((result) => {
            if (result.actions.length || result.intents.length) {
              setActionsMenu(result);
            }
          }).catch((error) => {
            logger.error("Failed to load select series actions", error);
          });
        }
      },
      DEBOUNCING_TIME,
      { leading: false, trailing: true }
    ),
    [logger]
  );
  useEffect(() => {
    if (!!actionsMenu.actions.length || !!actionsMenu.intents.length) {
      setActionsMenu(emptyActionsMenu);
    }
    props.actions && fetchActions(props.actions);
    return () => {
      fetchActions.cancel();
    };
  }, [props.range[0], props.range[1]]);
  const hasActions = !!actionsMenu.actions.length;
  const hasIntents = !!actionsMenu.intents.length;
  return /* @__PURE__ */ jsxs(
    OverlayTooltip,
    {
      isInteractive: props.selectFinished,
      absoluteX: props.position.x,
      absoluteY: props.position.y,
      children: [
        props.header,
        (!!props.children || hasActions || hasIntents) && /* @__PURE__ */ jsxs(Fragment, { children: [
          /* @__PURE__ */ jsx(
            Divider,
            {
              orientation: "horizontal",
              className: divider
            },
            "headerDivider"
          ),
          /* @__PURE__ */ jsxs(
            "div",
            {
              style: {
                maxHeight: "96px",
                overflowY: "auto",
                overflowX: "hidden"
              },
              children: [
                props.children,
                hasActions ? /* @__PURE__ */ jsx(CustomActions, { actions: actionsMenu.actions }) : null,
                hasIntents && (hasActions || !!props.children) ? /* @__PURE__ */ jsx(
                  Divider,
                  {
                    orientation: "horizontal",
                    className: divider
                  },
                  "actionsDivider"
                ) : null,
                hasIntents ? /* @__PURE__ */ jsx(CustomIntents, { intents: actionsMenu?.intents }) : null
              ]
            }
          )
        ] })
      ]
    }
  );
};
export {
  SelectionTooltipWithAsyncActions
};
