// @dynatrace/strato-components/esm/charts/core/components/selection-area/Selection.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { isUndefined } from "lodash-es";
import { useRef, useState } from "react";
import { SelectionEdge } from "./components/SelectionEdge.js";
import { SelectionBackdrop } from "./SelectionBackdrop.js";
import { changedEdge, LEFT_EDGE_ID, RIGHT_EDGE_ID } from "./utils.js";
import { KNOB_HEIGHT } from "../overlays/constants.js";
import { clampSelectionArea } from "./utils/clamp-selection-area.js";
import { SELECTION_RIGHT_OVERFLOW } from "../interactions-overlay/constants.js";
import { normalizeSelectionArea } from "../interactions-overlay/utils/normalize-selection-area.js";
const extendX = (relativeAbsolute, x) => ({
  ...relativeAbsolute,
  absolute: {
    ...relativeAbsolute.absolute,
    x: relativeAbsolute.absolute.x + x
  },
  relative: {
    ...relativeAbsolute.relative,
    x: relativeAbsolute.relative.x + x
  }
});
const Selection = (props) => {
  const {
    startX,
    endX,
    height,
    width,
    focusable,
    interactable,
    bottomFormatter,
    topFormatter,
    bounds,
    handleSelectionAreaStartKeyDown,
    handleSelectionAreaEndKeyDown,
    start,
    end,
    onSelectionChange,
    isBottomLabelRotated = false,
    isTopLabelRotated = false
  } = props;
  const range = [0, width + SELECTION_RIGHT_OVERFLOW];
  const selectionEdgePivotRef = useRef(void 0);
  const [hoveredEdge, setHoveredEdge] = useState("");
  const onDragMove = (x) => {
    if (!isUndefined(selectionEdgePivotRef.current)) {
      const position = {
        relative: {
          x: x - bounds.left,
          y: height / 2
        },
        absolute: {
          y: height / 2 + bounds.top,
          x
        }
      };
      const nextSelection = clampSelectionArea(
        normalizeSelectionArea([position, selectionEdgePivotRef.current]),
        range
      );
      const edge = changedEdge([startX, endX], nextSelection);
      if (edge) {
        setHoveredEdge(edge);
      }
      onSelectionChange?.({ selection: nextSelection, accessible: false });
    }
  };
  const onWholeSelectionMove = (x) => {
    const newSelectionAreaStart = extendX(start, x);
    const newSelectionAreaEnd = extendX(end, x);
    const nextSelection = clampSelectionArea(
      normalizeSelectionArea([newSelectionAreaStart, newSelectionAreaEnd]),
      range
    );
    onSelectionChange?.({ selection: nextSelection, accessible: true });
  };
  const edgesConfig = [
    {
      id: LEFT_EDGE_ID,
      x: startX,
      onKeyDown: handleSelectionAreaStartKeyDown,
      pivot: end
    },
    {
      id: RIGHT_EDGE_ID,
      x: endX,
      onKeyDown: handleSelectionAreaEndKeyDown,
      pivot: start
    }
  ];
  const formatters = {
    top: topFormatter || ((_x) => void 0),
    bottom: bottomFormatter || ((_x) => void 0)
  };
  const y = height / 2 - KNOB_HEIGHT / 2;
  const pointerHanders = (id) => ({
    onPointerEnter: () => {
      setHoveredEdge(id);
    },
    onPointerLeave: () => {
      setHoveredEdge("");
    }
  });
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(
      SelectionBackdrop,
      {
        focusable,
        startX,
        endX,
        height,
        isSynchronizedSelection: false,
        interactable,
        onDragMove: onWholeSelectionMove,
        hasTopLabels: !isUndefined(topFormatter),
        hasBottomLabels: !isUndefined(bottomFormatter)
      }
    ),
    edgesConfig.map((edge) => {
      return /* @__PURE__ */ jsx(
        SelectionEdge,
        {
          edgeId: edge.id,
          hoveredEdge,
          x: edge.x,
          y,
          focusable,
          interactable,
          onKeyDown: edge.onKeyDown,
          onDragMove,
          onDragStart: () => {
            selectionEdgePivotRef.current = edge.pivot;
          },
          onDragEnd: () => selectionEdgePivotRef.current = void 0,
          topLabel: formatters.top(edge.x),
          bottomLabel: formatters.bottom(edge.x),
          chartBounds: bounds,
          isBottomLabelRotated,
          isTopLabelRotated,
          pointerHandlers: pointerHanders(edge.id)
        },
        edge.id
      );
    }),
    hoveredEdge === LEFT_EDGE_ID && /* @__PURE__ */ jsx(
      SelectionEdge,
      {
        edgeId: edgesConfig[0].id,
        hoveredEdge,
        x: edgesConfig[0].x,
        y,
        focusable,
        interactable: false,
        onKeyDown: edgesConfig[0].onKeyDown,
        onDragMove,
        onDragStart: () => {
          selectionEdgePivotRef.current = edgesConfig[0].pivot;
        },
        onDragEnd: () => selectionEdgePivotRef.current = void 0,
        topLabel: formatters.top(edgesConfig[0].x),
        bottomLabel: formatters.bottom(edgesConfig[0].x),
        chartBounds: bounds,
        isBottomLabelRotated,
        isTopLabelRotated,
        pointerHandlers: pointerHanders(edgesConfig[0].id)
      },
      edgesConfig[0].id
    )
  ] });
};
export {
  Selection
};
