// @dynatrace/strato-components/esm/charts/core/components/selection-area/ZoomSelection.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { useImperativeHandle, useRef } from "react";
import { VerticalCrosshair } from "../crosshair/VerticalCrosshair.js";
import { SELECTION_EDGE_THRESHOLD } from "../interactions-overlay/constants.js";
import { ZOOM_AREA_STYLES } from "../overlays/constants.js";
import { useSelectionAreaPathRef } from "./hooks/use-selection-area-path-ref.js";
const ZoomSelection = ({
  startX,
  endX,
  height,
  focusable,
  interactable,
  bottomFormatter,
  topFormatter,
  bounds,
  handleSelectionAreaKeyDown,
  handleSelectionAreaStartKeyDown,
  handleSelectionAreaEndKeyDown,
  isBottomLabelRotated = false,
  isTopLabelRotated = false
}) => {
  const areaStyles = ZOOM_AREA_STYLES;
  const selectionAreaRef = useRef(null);
  const forwardedRef = useSelectionAreaPathRef();
  useImperativeHandle(
    forwardedRef,
    () => ({
      element: selectionAreaRef.current,
      focus() {
        selectionAreaRef.current?.focus();
      }
    }),
    []
  );
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(
      "rect",
      {
        ref: selectionAreaRef,
        x: startX,
        y: 0,
        width: endX - startX,
        height,
        tabIndex: focusable ? 0 : -1,
        onKeyDown: handleSelectionAreaKeyDown,
        ...areaStyles,
        rx: "0"
      }
    ),
    /* @__PURE__ */ jsx(
      "rect",
      {
        x: startX - SELECTION_EDGE_THRESHOLD,
        y: 0,
        width: SELECTION_EDGE_THRESHOLD * 5,
        height,
        fill: "transparent",
        onMouseDown: () => false,
        style: {
          cursor: "col-resize",
          pointerEvents: interactable ? "all" : "none"
        }
      }
    ),
    /* @__PURE__ */ jsx(
      VerticalCrosshair,
      {
        x: startX,
        bottomLabel: bottomFormatter ? bottomFormatter(startX) : void 0,
        topLabel: topFormatter ? topFormatter(startX) : void 0,
        height,
        dashed: false,
        focusable,
        chartBounds: bounds,
        onKeyDown: handleSelectionAreaStartKeyDown,
        variant: "zoom",
        isTopLabelRotated,
        isBottomLabelRotated
      }
    ),
    /* @__PURE__ */ jsx(
      "rect",
      {
        x: endX - SELECTION_EDGE_THRESHOLD,
        y: 0,
        width: SELECTION_EDGE_THRESHOLD * 5,
        height,
        fill: "transparent",
        style: {
          cursor: "col-resize",
          pointerEvents: interactable ? "all" : "none"
        }
      }
    ),
    /* @__PURE__ */ jsx(
      VerticalCrosshair,
      {
        x: endX,
        bottomLabel: bottomFormatter ? bottomFormatter(endX) : void 0,
        topLabel: topFormatter ? topFormatter(endX) : void 0,
        height,
        dashed: false,
        focusable,
        chartBounds: bounds,
        onKeyDown: handleSelectionAreaEndKeyDown,
        variant: "zoom",
        isTopLabelRotated,
        isBottomLabelRotated
      }
    )
  ] });
};
export {
  ZoomSelection
};
