// @dynatrace/strato-components/charts/xy-chart-adoption/histogram/HistogramToXY.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var HistogramToXY_exports = {};
__export(HistogramToXY_exports, {
  HistogramToXY: () => HistogramToXY
});
module.exports = __toCommonJS(HistogramToXY_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_use_histogram_to_xy_color_rules = require("./hooks/use-histogram-to-xy-color-rules.js");
var import_transformer_annotations_to_xy = require("./transformer-annotations-to-xy.js");
var import_transformer_data_to_xy = require("./transformer-data-to-xy.js");
var import_transformer_selection_to_histogram = require("./transformer-selection-to-histogram.js");
var import_transformer_series_actions_to_xy = require("./transformer-series-actions-to-xy.js");
var import_transformer_thresholds_to_xy = require("./transformer-thresholds-to-xy.js");
var import_axis_props_mapping = require("./utils/axis-props-mapping.js");
var import_build_histogram_imperative_handle = require("./utils/build-histogram-imperative-handle.js");
var import_build_y_axes_config = require("./utils/build-y-axes-config.js");
var import_define_axis_id = require("./utils/define-axis-id.js");
var import_derive_xychart_props_from_config = require("./utils/derive-xychart-props-from-config.js");
var import_format_csv = require("./utils/format-csv.js");
var import_download = require("../../../core/utils/download.js");
var import_get_chart_states_template = require("../../core/components/chart-states/utils/get-chart-states-template.js");
var import_constants = require("../../core/constants.js");
var import_useHistogramChartConfig = require("../../histogram/hooks/useHistogramChartConfig.js");
var import_XYChartRoot = require("../../xy-chart/XYChartRoot.js");
const getAnnotationField = (d, key) => d[key];
const HistogramToXY = (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    data: propsData,
    loading,
    children,
    className,
    style,
    width,
    height,
    colorPalette,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    truncationMode,
    valueRepresentation,
    seriesActions,
    initialZoom,
    currentZoom,
    onZoomChange,
    infiniteZoom
  } = props;
  const config = (0, import_useHistogramChartConfig.useHistogramChartConfig)(children, {
    colorPalette,
    truncationMode,
    valueRepresentation,
    initialZoom,
    currentZoom,
    onZoomChange,
    infiniteZoom
  });
  const { barSeriesProps, units, seriesColors } = (0, import_transformer_data_to_xy.transformHistogramDataToXYChart)(propsData);
  const allColorRules = (0, import_use_histogram_to_xy_color_rules.useHistogramToXYColorRules)({
    data: propsData,
    seriesColors,
    colorRules: config.colorRules
  });
  const { emptyState, errorState } = (0, import_get_chart_states_template.getChartStatesTemplates)(children);
  const downloadDataCallback = (0, import_react.useCallback)(() => {
    const csvString = (0, import_format_csv.formatCSV)(propsData);
    (0, import_download.downloadCSV)(csvString, import_constants.CSS_DOWNLOAD_FILENAME);
  }, [propsData]);
  const selectActions = config.select?.actions;
  const mappedSeriesActions = (0, import_react.useMemo)(
    () => seriesActions ? (0, import_transformer_series_actions_to_xy.mapHistogramSeriesActionsToXY)(seriesActions, propsData) : void 0,
    [seriesActions, propsData]
  );
  const yAxesConfig = (0, import_build_y_axes_config.buildYAxesConfigFromConfig)(config);
  const mappedThresholds = (0, import_react.useMemo)(
    () => (0, import_transformer_thresholds_to_xy.mapHistogramThresholdsToXY)(config.thresholds, [...units]),
    [config.thresholds, units]
  );
  const {
    xAxisConfig,
    disableZoom,
    disablePan,
    toolbarEnabled,
    toolbarPlacement,
    toolbarIntents,
    onZoomChange: effectiveOnZoomChange
  } = (0, import_derive_xychart_props_from_config.deriveXYChartPropsFromConfig)(config, onZoomChange);
  const xyRef = (0, import_react.useRef)(null);
  (0, import_react.useImperativeHandle)(
    forwardedRef,
    () => (0, import_build_histogram_imperative_handle.buildHistogramImperativeHandle)(xyRef, config),
    [config]
  );
  const annotations = config.annotations;
  const xAxisId = "x-axis";
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_XYChartRoot.XYChart,
    {
      data: [],
      ref: xyRef,
      loading,
      className,
      style,
      width,
      height,
      colorPalette: config.colorPalette,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      truncationMode: config.truncationMode,
      __downloadDataCallback: downloadDataCallback,
      children: [
        errorState && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartRoot.XYChart.ErrorState, { children: errorState }),
        emptyState && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartRoot.XYChart.EmptyState, { children: emptyState }),
        selectActions ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_XYChartRoot.XYChart.Select,
          {
            actions: (selectedSeries, selectionDomain) => selectActions(
              (0, import_transformer_selection_to_histogram.mapSelectedSeriesToHistogram)(selectedSeries),
              (0, import_transformer_selection_to_histogram.mapSelectionDomainToHistogramDomains)(selectionDomain)
            )
          }
        ) : null,
        barSeriesProps.map(({ data, unit }, index) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_XYChartRoot.XYChart.BarSeries,
          {
            data,
            xAxisId,
            yAxisId: (0, import_define_axis_id.defineAxisId)(unit),
            x0Accessor: "x0",
            x1Accessor: "x1",
            y1Accessor: "y1",
            variant: config.valueRepresentation === "relative" && propsData.length === 1 ? "single" : "stack",
            seriesIdAccessor: "seriesId",
            nameAccessor: "name",
            actions: mappedSeriesActions
          },
          index
        )),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_XYChartRoot.XYChart.XAxis,
          {
            id: xAxisId,
            type: xAxisConfig.scale === "log" ? "log" : "numerical",
            position: "bottom",
            disableZoom,
            disablePan,
            initialZoom: config.initialZoom,
            currentZoom: config.currentZoom,
            onZoomChange: effectiveOnZoomChange,
            infiniteZoom: config.infiniteZoom,
            ...xAxisConfig
          }
        ),
        [...units].map((unit, index) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_XYChartRoot.XYChart.YAxis,
          {
            ...(0, import_axis_props_mapping.buildYAxisProps)(
              index,
              unit,
              yAxesConfig,
              config.valueRepresentation
            )
          },
          (0, import_define_axis_id.defineAxisId)(unit)
        )),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartRoot.XYChart.Legend, { ...config.legend }),
        allColorRules.map((rule) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_XYChartRoot.XYChart.ColorRule,
          {
            ...rule
          },
          `${rule.comparator}-${rule.matchValue}-${rule.valueAccessor}`
        )),
        mappedThresholds.map((threshold) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_XYChartRoot.XYChart.ThresholdIndicator,
          {
            ...threshold
          },
          `${threshold.yAxisId}-${threshold.color}-${threshold.label ?? ""}`
        )),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartRoot.XYChart.Toolbar, { hidden: !toolbarEnabled, placement: toolbarPlacement }),
        toolbarIntents?.map((intentProps, index) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartRoot.XYChart.Intent, { ...intentProps }, index)),
        annotations && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
          import_XYChartRoot.XYChart.Annotations,
          {
            data: (0, import_transformer_annotations_to_xy.transformHistogramAnnotationsToXY)(annotations.tracks),
            xAxisId,
            showLabels: annotations.showLabels,
            fixedTracks: annotations.visibleTracksLimit,
            annotationsActions: annotations.annotationsActions,
            children: [
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_XYChartRoot.XYChartAnnotations.Track,
                {
                  trackIdAccessor: import_transformer_annotations_to_xy.ANNOTATIONS_KEYS.trackId,
                  labelAccessor: import_transformer_annotations_to_xy.ANNOTATIONS_KEYS.trackLabel,
                  colorAccessor: import_transformer_annotations_to_xy.ANNOTATIONS_KEYS.trackColor,
                  hidden: (d) => getAnnotationField(d, import_transformer_annotations_to_xy.ANNOTATIONS_KEYS.trackHidden),
                  symbol: (d) => getAnnotationField(d, import_transformer_annotations_to_xy.ANNOTATIONS_KEYS.trackSymbol),
                  indicatorsDisplay: (d) => getAnnotationField(
                    d,
                    import_transformer_annotations_to_xy.ANNOTATIONS_KEYS.trackIndicators
                  )
                }
              ),
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_XYChartRoot.XYChartAnnotations.Marker,
                {
                  startAccessor: import_transformer_annotations_to_xy.ANNOTATIONS_KEYS.markerStart,
                  endAccessor: import_transformer_annotations_to_xy.ANNOTATIONS_KEYS.markerEnd,
                  trackByAccessor: import_transformer_annotations_to_xy.ANNOTATIONS_KEYS.trackId,
                  titleAccessor: import_transformer_annotations_to_xy.ANNOTATIONS_KEYS.markerTitle,
                  descriptionAccessor: import_transformer_annotations_to_xy.ANNOTATIONS_KEYS.markerDescription,
                  colorAccessor: import_transformer_annotations_to_xy.ANNOTATIONS_KEYS.markerColor,
                  hidden: (d) => getAnnotationField(d, import_transformer_annotations_to_xy.ANNOTATIONS_KEYS.markerHidden),
                  symbol: (d) => getAnnotationField(d, import_transformer_annotations_to_xy.ANNOTATIONS_KEYS.markerSymbol),
                  indicatorsDisplay: (d) => getAnnotationField(
                    d,
                    import_transformer_annotations_to_xy.ANNOTATIONS_KEYS.markerIndicators
                  ),
                  priority: (d) => getAnnotationField(d, import_transformer_annotations_to_xy.ANNOTATIONS_KEYS.markerPriority)
                }
              ),
              annotations.tooltip?.template && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_XYChartRoot.XYChartAnnotations.Tooltip, { hidden: annotations.tooltip.hidden, children: annotations.tooltip.template })
            ]
          }
        )
      ]
    }
  );
});
HistogramToXY.displayName = "HistogramToXY";
