// @dynatrace/strato-components/charts/xy-chart-adoption/histogram/transformer-annotations-to-xy.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var transformer_annotations_to_xy_exports = {};
__export(transformer_annotations_to_xy_exports, {
  ANNOTATIONS_KEYS: () => ANNOTATIONS_KEYS,
  transformHistogramAnnotationsToXY: () => transformHistogramAnnotationsToXY
});
module.exports = __toCommonJS(transformer_annotations_to_xy_exports);
const ANNOTATIONS_KEYS = {
  trackId: "__annotationsTrackId",
  trackLabel: "__annotationsTrackLabel",
  trackColor: "__annotationsTrackColor",
  trackHidden: "__annotationsTrackHidden",
  trackSymbol: "__annotationsTrackSymbol",
  trackIndicators: "__annotationsTrackIndicators",
  markerStart: "__annotationsMarkerStart",
  markerEnd: "__annotationsMarkerEnd",
  markerTitle: "__annotationsMarkerTitle",
  markerDescription: "__annotationsMarkerDescription",
  markerColor: "__annotationsMarkerColor",
  markerHidden: "__annotationsMarkerHidden",
  markerSymbol: "__annotationsMarkerSymbol",
  markerIndicators: "__annotationsMarkerIndicators",
  markerPriority: "__annotationsMarkerPriority"
};
const transformHistogramAnnotationsToXY = (tracks) => {
  const data = [];
  tracks.forEach((track, trackIndex) => {
    const histogramTrack = track;
    const trackId = `track-${trackIndex}`;
    track.markers.forEach((marker) => {
      const histogramMarker = marker;
      data.push({
        [ANNOTATIONS_KEYS.trackId]: trackId,
        [ANNOTATIONS_KEYS.trackLabel]: histogramTrack.label,
        [ANNOTATIONS_KEYS.trackColor]: histogramTrack.color,
        [ANNOTATIONS_KEYS.trackHidden]: histogramTrack.hidden,
        [ANNOTATIONS_KEYS.trackSymbol]: histogramTrack.symbol,
        [ANNOTATIONS_KEYS.trackIndicators]: histogramTrack.indicatorsDisplay,
        [ANNOTATIONS_KEYS.markerStart]: histogramMarker.start,
        [ANNOTATIONS_KEYS.markerEnd]: histogramMarker.end,
        [ANNOTATIONS_KEYS.markerTitle]: histogramMarker.title,
        [ANNOTATIONS_KEYS.markerDescription]: histogramMarker.description,
        [ANNOTATIONS_KEYS.markerColor]: histogramMarker.color,
        [ANNOTATIONS_KEYS.markerHidden]: histogramMarker.hidden,
        [ANNOTATIONS_KEYS.markerSymbol]: histogramMarker.symbol,
        [ANNOTATIONS_KEYS.markerIndicators]: histogramMarker.indicatorsDisplay,
        [ANNOTATIONS_KEYS.markerPriority]: histogramMarker.priority
      });
    });
  });
  return data;
};
