// @dynatrace/strato-components/charts/xy-chart-adoption/histogram/transformer-data-to-xy.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var transformer_data_to_xy_exports = {};
__export(transformer_data_to_xy_exports, {
  transformHistogramDataToXYChart: () => transformHistogramDataToXYChart
});
module.exports = __toCommonJS(transformer_data_to_xy_exports);
function transformHistogramDataToXYChart(data) {
  const seriesColors = [];
  const units = /* @__PURE__ */ new Set();
  const unitToData = /* @__PURE__ */ new Map();
  const unitToSeriesCount = /* @__PURE__ */ new Map();
  data.forEach((series, index) => {
    const { bins, name, color, unit, ...extraSeriesProps } = series;
    if (color) {
      seriesColors.push({ name, color });
    }
    const binData = bins.map((bin) => ({
      ...extraSeriesProps,
      x0: bin.from,
      x1: bin.to,
      y1: bin.value,
      name,
      seriesId: index.toString()
    }));
    unitToData.set(unit, [...unitToData.get(unit) ?? [], ...binData]);
    unitToSeriesCount.set(unit, (unitToSeriesCount.get(unit) ?? 0) + 1);
    units.add(unit);
  });
  return {
    barSeriesProps: Array.from(unitToData, ([unit, data2]) => ({
      unit,
      data: data2,
      seriesCount: unitToSeriesCount.get(unit) ?? 1
    })),
    units,
    seriesColors
  };
}
