// @dynatrace-sdk/react-hooks/cjs/index.js
/**
 * @license
 * Copyright 2023-2026 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/util/react-hooks/src/index.ts
var src_exports = {};
__export(src_exports, {
  DqlQueryParamsContext: () => DqlQueryParamsContext,
  DqlQueryParamsProvider: () => DqlQueryParamsProvider,
  UseUsersProvider: () => UseUsersProvider,
  _UseDqlQueryClientContext: () => _UseDqlQueryClientContext,
  _UseDqlQueryClientProvider: () => _UseDqlQueryClientProvider,
  getGrailFieldsQueryOptions: () => getGrailFieldsQueryOptions,
  useAccessorPermissionsV2: () => useAccessorPermissionsV2,
  useAllUsersPermissionsV2: () => useAllUsersPermissionsV2,
  useAnalyzer: () => useAnalyzer,
  useAppFunction: () => useAppFunction,
  useAppState: () => useAppState,
  useAppStates: () => useAppStates,
  useCreateDocument: () => useCreateDocument,
  useCreatePermissionsV2: () => useCreatePermissionsV2,
  useCreateSettings: () => useCreateSettings,
  useCreateSettingsV2: () => useCreateSettingsV2,
  useDeleteAccessorPermissionsV2: () => useDeleteAccessorPermissionsV2,
  useDeleteAllUsersPermissionsV2: () => useDeleteAllUsersPermissionsV2,
  useDeleteAppState: () => useDeleteAppState,
  useDeleteAppStates: () => useDeleteAppStates,
  useDeleteDocument: () => useDeleteDocument,
  useDeleteSettings: () => useDeleteSettings,
  useDeleteSettingsV2: () => useDeleteSettingsV2,
  useDeleteUserAppState: () => useDeleteUserAppState,
  useDeleteUserAppStates: () => useDeleteUserAppStates,
  useDocument: () => useDocument,
  useDocumentMetaData: () => useDocumentMetaData,
  useDownloadDocument: () => useDownloadDocument,
  useDql: () => useDql,
  useDqlQuery: () => useDqlQuery,
  useEffectivePermissions: () => useEffectivePermissions,
  useEffectivePermissionsV2: () => useEffectivePermissionsV2,
  useGrailFields: () => useGrailFields,
  useListDocuments: () => useListDocuments,
  usePermissionsV2: () => usePermissionsV2,
  useSetAppState: () => useSetAppState,
  useSetUserAppState: () => useSetUserAppState,
  useSettings: () => useSettings,
  useSettingsObjects: () => useSettingsObjects,
  useSettingsObjectsV2: () => useSettingsObjectsV2,
  useSettingsV2: () => useSettingsV2,
  useTransferOwnershipV2: () => useTransferOwnershipV2,
  useUpdateAccessorPermissionsV2: () => useUpdateAccessorPermissionsV2,
  useUpdateAllUsersPermissionsV2: () => useUpdateAllUsersPermissionsV2,
  useUpdateDocument: () => useUpdateDocument,
  useUpdateDocumentMetadata: () => useUpdateDocumentMetadata,
  useUpdateSettings: () => useUpdateSettings,
  useUpdateSettingsV2: () => useUpdateSettingsV2,
  useUser: () => useUser,
  useUserAppState: () => useUserAppState,
  useUserAppStates: () => useUserAppStates,
  useUsers: () => useUsers
});
module.exports = __toCommonJS(src_exports);

// packages/util/react-hooks/src/useAnalyzer.ts
var import_client_davis_analyzers = require("@dynatrace-sdk/client-davis-analyzers");

// packages/util/react-hooks/src/react-async-hook/react-async-hook.ts
var import_react = require("react");
var useIsomorphicLayoutEffect = typeof window !== "undefined" && typeof window.document !== "undefined" && typeof window.document.createElement !== "undefined" ? import_react.useLayoutEffect : import_react.useEffect;
var useGetter = (t) => {
  const ref = (0, import_react.useRef)(t);
  useIsomorphicLayoutEffect(() => {
    ref.current = t;
  });
  return (0, import_react.useCallback)(() => ref.current, [ref]);
};
var InitialAsyncState = {
  status: "not-requested",
  loading: false,
  result: void 0,
  error: void 0
};
var InitialAsyncLoadingState = {
  status: "loading",
  loading: true,
  result: void 0,
  error: void 0
};
var defaultSetLoading = (_asyncState) => InitialAsyncLoadingState;
var defaultSetResult = (result, _asyncState) => ({
  status: "success",
  loading: false,
  result,
  error: void 0
});
var defaultSetError = (error, _asyncState) => ({
  status: "error",
  loading: false,
  result: void 0,
  error
});
var noop = () => {
};
var DefaultOptions = {
  initialState: (options) => options && options.executeOnMount ? InitialAsyncLoadingState : InitialAsyncState,
  executeOnMount: true,
  executeOnUpdate: true,
  setLoading: defaultSetLoading,
  setResult: defaultSetResult,
  setError: defaultSetError,
  onSuccess: noop,
  onError: noop
};
var normalizeOptions = (options) => ({
  ...DefaultOptions,
  ...options
});
var useAsyncState = (options) => {
  const [value, setValue] = (0, import_react.useState)(() => options.initialState(options));
  const reset = (0, import_react.useCallback)(() => setValue(options.initialState(options)), [setValue, options]);
  const setLoading = (0, import_react.useCallback)(() => setValue(options.setLoading(value)), [value, setValue]);
  const setResult = (0, import_react.useCallback)(
    (result) => setValue(options.setResult(result, value)),
    [value, setValue]
  );
  const setError = (0, import_react.useCallback)(
    (error) => setValue(options.setError(error, value)),
    [value, setValue]
  );
  const merge = (0, import_react.useCallback)(
    (state) => setValue({
      ...value,
      ...state
    }),
    [value, setValue]
  );
  return {
    value,
    set: setValue,
    merge,
    reset,
    setLoading,
    setResult,
    setError
  };
};
var useIsMounted = () => {
  const ref = (0, import_react.useRef)(false);
  (0, import_react.useEffect)(() => {
    ref.current = true;
    return () => {
      ref.current = false;
    };
  }, []);
  return () => ref.current;
};
var useCurrentPromise = () => {
  const ref = (0, import_react.useRef)(null);
  return {
    set: (promise) => ref.current = promise,
    get: () => ref.current,
    is: (promise) => ref.current === promise
  };
};
var useAsyncInternal = (asyncFunction, params, options) => {
  !params && (params = []);
  const normalizedOptions = normalizeOptions(options);
  const [currentParams, setCurrentParams] = (0, import_react.useState)(null);
  const AsyncState = useAsyncState(normalizedOptions);
  const isMounted = useIsMounted();
  const CurrentPromise = useCurrentPromise();
  const shouldHandlePromise = (p) => isMounted() && CurrentPromise.is(p);
  const executeAsyncOperation = (...args) => {
    const promise = (async () => asyncFunction(...args))();
    setCurrentParams(args);
    CurrentPromise.set(promise);
    AsyncState.setLoading();
    promise.then(
      (result) => {
        if (shouldHandlePromise(promise)) {
          AsyncState.setResult(result);
        }
        normalizedOptions.onSuccess(result, {
          isCurrent: () => CurrentPromise.is(promise)
        });
      },
      (error) => {
        if (shouldHandlePromise(promise)) {
          AsyncState.setError(error);
        }
        normalizedOptions.onError(error, {
          isCurrent: () => CurrentPromise.is(promise)
        });
      }
    );
    return promise;
  };
  const getLatestExecuteAsyncOperation = useGetter(executeAsyncOperation);
  const executeAsyncOperationMemo = (0, import_react.useCallback)(
    (...args) => {
      try {
        return getLatestExecuteAsyncOperation()(...args);
      } catch (e) {
        return Promise.resolve(void 0);
      }
    },
    [getLatestExecuteAsyncOperation]
  );
  const isMounting = !isMounted();
  (0, import_react.useEffect)(() => {
    const execute = () => getLatestExecuteAsyncOperation()(...params);
    isMounting && normalizedOptions.executeOnMount && execute();
    !isMounting && normalizedOptions.executeOnUpdate && execute();
  }, params);
  return {
    ...AsyncState.value,
    set: AsyncState.set,
    merge: AsyncState.merge,
    reset: AsyncState.reset,
    execute: executeAsyncOperationMemo,
    currentPromise: CurrentPromise.get(),
    currentParams
  };
};
function useAsync(asyncFunction, params, options) {
  return useAsyncInternal(asyncFunction, params, options);
}
var useAsyncAbortable = (asyncFunction, params, options) => {
  const abortControllerRef = (0, import_react.useRef)();
  const asyncFunctionWrapper = async function() {
    if (abortControllerRef.current) {
      abortControllerRef.current.abort();
    }
    const abortController = new AbortController();
    abortControllerRef.current = abortController;
    try {
      return await asyncFunction(abortController.signal, ...arguments);
    } finally {
      if (abortControllerRef.current === abortController) {
        abortControllerRef.current = void 0;
      }
    }
  };
  return {
    ...useAsync(asyncFunctionWrapper, params, options),
    cancel: () => abortControllerRef?.current?.abort()
  };
};
var useAsyncCallback = (asyncFunction, options) => {
  return useAsyncInternal(
    asyncFunction,
    // Hacky but in such case we don't need the params,
    // because async function is only executed manually
    [],
    {
      ...options,
      executeOnMount: false,
      executeOnUpdate: false
    }
  );
};

// packages/util/react-hooks/src/util/delay-util.ts
var delay = (ms, value, signal) => new Promise((resolve, reject) => {
  const listener = () => {
    clearTimeout(timer);
    reject(value);
  };
  const timer = setTimeout(() => {
    signal?.removeEventListener("abort", listener);
    resolve(value);
  }, ms);
  signal?.addEventListener("abort", listener);
});

// packages/util/react-hooks/src/useAnalyzer.ts
var useAnalyzer = (params, options = { autoFetch: true, autoFetchOnUpdate: true }) => {
  const onError = typeof options.onError === "function" ? options.onError : () => void 0;
  const result = useAsyncAbortable(
    (signal) => doExecute(params, signal),
    [JSON.stringify(params)],
    {
      executeOnMount: options.autoFetch,
      executeOnUpdate: options.autoFetchOnUpdate,
      onError
    }
  );
  return getResult(result);
};
var doExecute = async (params, abortSignal) => {
  const executeResult = await import_client_davis_analyzers.analyzersClient.executeAnalyzer({ ...params, abortSignal });
  if (executeResult.result.executionStatus === "COMPLETED") {
    return executeResult.result;
  }
  let pollResult = null;
  while (executeResult.requestToken && (pollResult === null || pollResult.result.executionStatus === "RUNNING")) {
    try {
      await delay(1e3, null, abortSignal);
      pollResult = await import_client_davis_analyzers.analyzersClient.pollAnalyzerExecution({
        analyzerName: params.analyzerName,
        requestToken: executeResult.requestToken,
        abortSignal
      });
    } finally {
      await abortAnalyzer(abortSignal, params.analyzerName, executeResult.requestToken);
    }
  }
  return pollResult?.result;
};
var getResult = (result) => {
  let error = result.error;
  if (result.error?.name === "AbortError")
    error = void 0;
  return {
    data: result.result,
    error: result.error,
    errorDetails: error?.body?.error,
    isError: error ? true : false,
    isLoading: result.loading,
    isSuccess: !error && !result.loading && result.result,
    status: result.status,
    refetch: result.execute,
    cancel: result.cancel
  };
};
var abortAnalyzer = async function(signal, analyzerName, requestToken) {
  if (signal?.aborted) {
    await import_client_davis_analyzers.analyzersClient.cancelAnalyzerExecution({
      analyzerName,
      requestToken
    });
    return Promise.reject();
  }
};

// packages/util/react-hooks/src/useAppState.ts
var import_client_state = require("@dynatrace-sdk/client-state");
var import_react2 = require("react");
function useAppState(params, options = { autoFetch: true, autoFetchOnUpdate: true }) {
  const onError = typeof options.onError === "function" ? options.onError : () => void 0;
  const result = useAsyncAbortable(
    (signal) => {
      return import_client_state.stateClient.getAppState({ ...params, abortSignal: signal });
    },
    [JSON.stringify(params)],
    {
      executeOnMount: options.autoFetch,
      executeOnUpdate: options.autoFetchOnUpdate,
      onError
    }
  );
  return getResult2(result);
}
function useAppStates(params, options = { autoFetch: true, autoFetchOnUpdate: true }) {
  const onError = typeof options.onError === "function" ? options.onError : () => void 0;
  const result = useAsyncAbortable(
    (signal) => import_client_state.stateClient.getAppStates({ ...params, abortSignal: signal }),
    [JSON.stringify(params)],
    {
      executeOnMount: options.autoFetch,
      executeOnUpdate: options.autoFetchOnUpdate,
      onError
    }
  );
  return getResult2(result);
}
function useUserAppState(params, options = { autoFetch: true, autoFetchOnUpdate: true }) {
  const onError = typeof options.onError === "function" ? options.onError : () => void 0;
  const result = useAsyncAbortable(
    (signal) => {
      return import_client_state.stateClient.getUserAppState({ ...params, abortSignal: signal });
    },
    [JSON.stringify(params)],
    {
      executeOnMount: options.autoFetch,
      executeOnUpdate: options.autoFetchOnUpdate,
      onError
    }
  );
  return getResult2(result);
}
function useUserAppStates(params, options = { autoFetch: true, autoFetchOnUpdate: true }) {
  const onError = typeof options.onError === "function" ? options.onError : () => void 0;
  const result = useAsyncAbortable(
    (signal) => import_client_state.stateClient.getUserAppStates({ ...params, abortSignal: signal }),
    [JSON.stringify(params)],
    {
      executeOnMount: options.autoFetch,
      executeOnUpdate: options.autoFetchOnUpdate,
      onError
    }
  );
  return getResult2(result);
}
function useDeleteAppState() {
  const execute = (0, import_react2.useCallback)(async (params) => {
    await result.execute(params);
  }, []);
  const result = useAsyncCallback((params) => import_client_state.stateClient.deleteAppState({ ...params }));
  return getCallbackResult(result, execute);
}
function useDeleteAppStates() {
  const result = useAsyncCallback(() => import_client_state.stateClient.deleteAppStates());
  return getCallbackResult(result, result.execute);
}
function useDeleteUserAppState() {
  const execute = (0, import_react2.useCallback)(async (params) => {
    await result.execute(params);
  }, []);
  const result = useAsyncCallback((params) => import_client_state.stateClient.deleteUserAppState({ ...params }));
  return getCallbackResult(result, execute);
}
function useDeleteUserAppStates() {
  const result = useAsyncCallback(() => import_client_state.stateClient.deleteUserAppStates());
  return getCallbackResult(result, result.execute);
}
function useSetAppState() {
  const execute = (0, import_react2.useCallback)(async (params) => {
    await result.execute(params);
  }, []);
  const result = useAsyncCallback(
    (params) => import_client_state.stateClient.setAppState({ ...params })
  );
  return getCallbackResult(result, execute);
}
function useSetUserAppState() {
  const execute = (0, import_react2.useCallback)(async (params) => {
    await result.execute(params);
  }, []);
  const result = useAsyncCallback(
    (params) => import_client_state.stateClient.setUserAppState({ ...params })
  );
  return getCallbackResult(result, execute);
}
function getCallbackResult(result, execute) {
  const error = result.error;
  return {
    execute,
    data: result.result,
    error: result.error,
    errorDetails: error?.body?.error,
    isError: error ? true : false,
    isSuccess: !error && !result.loading && result.result,
    isLoading: result.loading,
    status: result.status
  };
}
function getResult2(result) {
  const error = result.error;
  return {
    data: result.result,
    error: result.error,
    errorDetails: error?.body?.error,
    isError: error ? true : false,
    isLoading: result.loading,
    isSuccess: !error && !result.loading && result.result,
    status: result.status,
    refetch: result.execute,
    cancel: result.cancel
  };
}

// packages/util/react-hooks/src/useDocument.ts
var import_client_document = require("@dynatrace-sdk/client-document");
var import_react3 = require("react");
function useDocument(params, options = { autoFetch: true, autoFetchOnUpdate: true }) {
  const onError = typeof options.onError === "function" ? options.onError : () => void 0;
  const result = useAsyncAbortable(
    (signal) => import_client_document.documentsClient.getDocument({ ...params, abortSignal: signal }),
    [JSON.stringify(params)],
    {
      executeOnMount: options.autoFetch,
      executeOnUpdate: options.autoFetchOnUpdate,
      onError
    }
  );
  return getResult3(result);
}
function useDocumentMetaData(params, options = { autoFetch: true, autoFetchOnUpdate: true }) {
  const onError = typeof options.onError === "function" ? options.onError : () => void 0;
  const result = useAsyncAbortable(
    (signal) => import_client_document.documentsClient.getDocumentMetadata({
      ...params,
      abortSignal: signal
    }),
    [JSON.stringify(params)],
    {
      executeOnMount: options.autoFetch,
      executeOnUpdate: options.autoFetchOnUpdate,
      onError
    }
  );
  return getResult3(result);
}
function useListDocuments(params, options = { autoFetch: true, autoFetchOnUpdate: true }) {
  const onError = typeof options.onError === "function" ? options.onError : () => void 0;
  const result = useAsyncAbortable(
    async (signal) => import_client_document.documentsClient.listDocuments({ ...params, abortSignal: signal }),
    [JSON.stringify(params)],
    {
      executeOnMount: options.autoFetch,
      executeOnUpdate: options.autoFetchOnUpdate,
      onError
    }
  );
  return getResult3(result);
}
function useDownloadDocument(params, options = { autoFetch: true, autoFetchOnUpdate: true }) {
  const onError = typeof options.onError === "function" ? options.onError : () => void 0;
  const result = useAsyncAbortable(
    (signal) => import_client_document.documentsClient.downloadDocumentContent({ ...params, abortSignal: signal }),
    [JSON.stringify(params)],
    {
      executeOnMount: options.autoFetch,
      executeOnUpdate: options.autoFetchOnUpdate,
      onError
    }
  );
  return getResult3(result);
}
function useCreateDocument() {
  const executeParams = (0, import_react3.useRef)();
  const execute = async (params) => {
    executeParams.current = params;
    await result.execute();
  };
  const result = useAsyncCallback(() => {
    if (executeParams.current) {
      return import_client_document.documentsClient.createDocument({ ...executeParams.current });
    }
  });
  return getCallbackResult2(result, execute);
}
function useUpdateDocument() {
  const executeParams = (0, import_react3.useRef)();
  const execute = async (params) => {
    executeParams.current = params;
    await result.execute();
  };
  const result = useAsyncCallback(() => {
    if (executeParams.current) {
      return import_client_document.documentsClient.updateDocument({ ...executeParams.current });
    }
  });
  return getCallbackResult2(result, execute);
}
function useUpdateDocumentMetadata() {
  const executeParams = (0, import_react3.useRef)();
  const execute = async (params) => {
    executeParams.current = params;
    await result.execute();
  };
  const result = useAsyncCallback(() => {
    if (executeParams.current) {
      return import_client_document.documentsClient.updateDocumentMetadata({ ...executeParams.current });
    }
  });
  return getCallbackResult2(result, execute);
}
function useDeleteDocument() {
  const executeParams = (0, import_react3.useRef)();
  const execute = async (params) => {
    executeParams.current = params;
    await result.execute();
  };
  const result = useAsyncCallback(() => {
    if (executeParams.current) {
      return import_client_document.documentsClient.deleteDocument({ ...executeParams.current });
    }
  });
  return getCallbackResult2(result, execute);
}
function getCallbackResult2(result, execute) {
  const error = result.error;
  return {
    execute,
    data: result.result,
    error: result.error,
    errorDetails: error?.body?.error,
    isError: error ? true : false,
    isSuccess: !error && !result.loading && result.result,
    isLoading: result.loading,
    status: result.status
  };
}
function getResult3(result) {
  const error = result.error;
  return {
    data: result.result,
    error: result.error,
    errorDetails: error?.body?.error,
    isError: error ? true : false,
    isLoading: result.loading,
    isSuccess: !error && !result.loading && result.result,
    status: result.status,
    refetch: result.execute,
    cancel: result.cancel
  };
}

// packages/util/react-hooks/src/useDqlQuery.ts
var import_client_query = require("@dynatrace-sdk/client-query");
function useDqlQuery(params, options = { autoFetch: true, autoFetchOnUpdate: true }) {
  const onError = typeof options.onError === "function" ? options.onError : () => void 0;
  const result = useAsyncAbortable(
    (signal) => doQuery(params, signal),
    [JSON.stringify(params)],
    {
      executeOnMount: options.autoFetch,
      executeOnUpdate: options.autoFetchOnUpdate,
      onError
    }
  );
  let error = result.error;
  if (result.error?.name === "AbortError")
    error = void 0;
  return {
    data: result.result,
    error: error ? result.error : void 0,
    errorDetails: error?.body?.error,
    isError: error ? true : false,
    isLoading: result.loading,
    isSuccess: !error && !result.loading && !!result.result,
    status: result.status,
    refetch: result.execute,
    cancel: result.cancel
  };
}
var doQuery = async function(params, signal) {
  const executeResult = await import_client_query.queryExecutionClient.queryExecute({
    ...params,
    abortSignal: signal
  });
  if (!(executeResult.state === import_client_query.QueryState.Running || executeResult.state === import_client_query.QueryState.NotStarted || executeResult.result === void 0)) {
    return executeResult.result;
  }
  let queryResult = null;
  while (executeResult.requestToken && (queryResult === null || queryResult.state === import_client_query.QueryState.Running || queryResult.state === import_client_query.QueryState.NotStarted)) {
    try {
      await delay(1e3, null, signal);
    } finally {
      await abortQuery(signal, executeResult.requestToken);
    }
    try {
      queryResult = await import_client_query.queryExecutionClient.queryPoll({
        requestToken: executeResult.requestToken,
        enrich: params?.enrich,
        abortSignal: signal
      });
    } finally {
      await abortQuery(signal, executeResult.requestToken);
    }
  }
  return queryResult?.result;
};
var abortQuery = async function(signal, requestToken) {
  if (signal?.aborted) {
    await import_client_query.queryExecutionClient.queryCancel({
      requestToken
    });
    return Promise.reject();
  }
};

// packages/util/react-hooks/src/useAppFunction.ts
var import_app_utils = require("@dynatrace-sdk/app-utils");
function useAppFunction({ name, data, responseType = "json" }, options = { autoFetch: true, autoFetchOnUpdate: true }) {
  const onError = typeof options.onError === "function" ? options.onError : () => void 0;
  const result = useAsyncAbortable(
    (signal) => import_app_utils.functions.call(name, { data, abortSignal: signal }).then((res) => res[responseType]()),
    [JSON.stringify(data)],
    {
      executeOnMount: options.autoFetch,
      executeOnUpdate: options.autoFetchOnUpdate,
      onError
    }
  );
  return {
    data: result.result,
    error: result.error,
    isError: !!result.error,
    isLoading: result.loading,
    isSuccess: !result.error && !result.loading && !!result.result,
    status: result.status,
    refetch: result.execute,
    cancel: result.cancel
  };
}

// packages/util/react-hooks/src/useSettings.ts
var import_client_app_settings = require("@dynatrace-sdk/client-app-settings");
var import_react4 = require("react");
function useSettings(params, options = { autoFetch: true, autoFetchOnUpdate: true }) {
  const onError = typeof options.onError === "function" ? options.onError : () => void 0;
  const result = useAsyncAbortable(
    (signal) => {
      return import_client_app_settings.appSettingsObjectsClient.getEffectiveAppSettingsValues({
        ...params,
        abortSignal: signal
      });
    },
    [JSON.stringify(params)],
    {
      executeOnMount: options.autoFetch,
      executeOnUpdate: options.autoFetchOnUpdate,
      onError
    }
  );
  return getResult4(result);
}
function useSettingsObjects(params, options = { autoFetch: true, autoFetchOnUpdate: true }) {
  const onError = typeof options.onError === "function" ? options.onError : () => void 0;
  const result = useAsyncAbortable(
    (signal) => {
      return import_client_app_settings.appSettingsObjectsClient.getAppSettingsObjects({
        ...params,
        abortSignal: signal
      });
    },
    [JSON.stringify(params)],
    {
      executeOnMount: options.autoFetch,
      executeOnUpdate: options.autoFetchOnUpdate,
      onError
    }
  );
  return getResult4(result);
}
function useEffectivePermissions(params, options = { autoFetch: true, autoFetchOnUpdate: true }) {
  const onError = typeof options.onError === "function" ? options.onError : () => void 0;
  const result = useAsyncAbortable(
    (signal) => {
      return import_client_app_settings.appSettingsObjectsClient.resolveEffectivePermissions({
        ...params,
        abortSignal: signal
      });
    },
    [JSON.stringify(params)],
    {
      executeOnMount: options.autoFetch,
      executeOnUpdate: options.autoFetchOnUpdate,
      onError
    }
  );
  return getResult4(result);
}
function useDeleteSettings() {
  const executeParams = (0, import_react4.useRef)();
  const execute = async (params) => {
    executeParams.current = params;
    await result.execute();
  };
  const result = useAsyncCallback(() => {
    if (executeParams.current) {
      return import_client_app_settings.appSettingsObjectsClient.deleteAppSettingsObjectByObjectId({
        ...executeParams.current
      });
    }
  });
  return getCallbackResult3(result, execute);
}
function useUpdateSettings() {
  const executeParams = (0, import_react4.useRef)();
  const execute = async (params) => {
    executeParams.current = params;
    await result.execute();
  };
  const result = useAsyncCallback(() => {
    if (executeParams.current) {
      return import_client_app_settings.appSettingsObjectsClient.putAppSettingsObjectByObjectId({ ...executeParams.current });
    }
  });
  return getCallbackResult3(result, execute);
}
function useCreateSettings() {
  const executeParams = (0, import_react4.useRef)();
  const execute = async (params) => {
    executeParams.current = params;
    await result.execute();
  };
  const result = useAsyncCallback(() => {
    if (executeParams.current) {
      return import_client_app_settings.appSettingsObjectsClient.postAppSettingsObject({ ...executeParams.current });
    }
  });
  return getCallbackResult3(result, execute);
}
function getCallbackResult3(result, execute) {
  const error = result.error;
  return {
    execute,
    data: result.result,
    error: error ? result.error : void 0,
    errorDetails: error?.body?.error,
    isError: error ? true : false,
    isSuccess: !error && !result.loading && result.result,
    isLoading: result.loading,
    status: result.status
  };
}
function getResult4(result) {
  const error = result.error;
  return {
    data: result.result,
    error: error ? result.error : void 0,
    errorDetails: error?.body?.error,
    isError: error ? true : false,
    isLoading: result.loading,
    isSuccess: !error && !result.loading && result.result,
    status: result.status,
    refetch: result.execute,
    cancel: result.cancel
  };
}

// packages/util/react-hooks/src/useSettingsV2.ts
var import_client_app_settings_v2 = require("@dynatrace-sdk/client-app-settings-v2");
var import_react5 = require("react");
function useSettingsV2(params, options = { autoFetch: true, autoFetchOnUpdate: true }) {
  const onError = typeof options.onError === "function" ? options.onError : () => void 0;
  const result = useAsyncAbortable(
    (signal) => {
      return import_client_app_settings_v2.appSettingsObjectsClient.getEffectiveAppSettingsValues({
        ...params,
        abortSignal: signal
      });
    },
    [JSON.stringify(params)],
    {
      executeOnMount: options.autoFetch,
      executeOnUpdate: options.autoFetchOnUpdate,
      onError
    }
  );
  return getResult5(result);
}
function useSettingsObjectsV2(params, options = { autoFetch: true, autoFetchOnUpdate: true }) {
  const onError = typeof options.onError === "function" ? options.onError : () => void 0;
  const result = useAsyncAbortable(
    (signal) => {
      return import_client_app_settings_v2.appSettingsObjectsClient.getAppSettingsObjects({
        ...params,
        abortSignal: signal
      });
    },
    [JSON.stringify(params)],
    {
      executeOnMount: options.autoFetch,
      executeOnUpdate: options.autoFetchOnUpdate,
      onError
    }
  );
  return getResult5(result);
}
function useEffectivePermissionsV2(params, options = { autoFetch: true, autoFetchOnUpdate: true }) {
  const onError = typeof options.onError === "function" ? options.onError : () => void 0;
  const result = useAsyncAbortable(
    (signal) => {
      return import_client_app_settings_v2.appSettingsObjectsClient.resolveEffectivePermissions({
        ...params,
        abortSignal: signal
      });
    },
    [JSON.stringify(params)],
    {
      executeOnMount: options.autoFetch,
      executeOnUpdate: options.autoFetchOnUpdate,
      onError
    }
  );
  return getResult5(result);
}
function useDeleteSettingsV2() {
  const executeParams = (0, import_react5.useRef)();
  const execute = async (params) => {
    executeParams.current = params;
    await result.execute();
  };
  const result = useAsyncCallback(() => {
    if (executeParams.current) {
      return import_client_app_settings_v2.appSettingsObjectsClient.deleteAppSettingsObjectByObjectId({
        ...executeParams.current
      });
    }
  });
  return getCallbackResult4(result, execute);
}
function useUpdateSettingsV2() {
  const executeParams = (0, import_react5.useRef)();
  const execute = async (params) => {
    executeParams.current = params;
    await result.execute();
  };
  const result = useAsyncCallback(() => {
    if (executeParams.current) {
      return import_client_app_settings_v2.appSettingsObjectsClient.putAppSettingsObjectByObjectId({
        ...executeParams.current
      });
    }
  });
  return getCallbackResult4(result, execute);
}
function useCreateSettingsV2() {
  const executeParams = (0, import_react5.useRef)();
  const execute = async (params) => {
    executeParams.current = params;
    await result.execute();
  };
  const result = useAsyncCallback(() => {
    if (executeParams.current) {
      return import_client_app_settings_v2.appSettingsObjectsClient.postAppSettingsObject({
        ...executeParams.current
      });
    }
  });
  return getCallbackResult4(result, execute);
}
function useTransferOwnershipV2() {
  const executeParams = (0, import_react5.useRef)();
  const execute = async (params) => {
    executeParams.current = params;
    await result.execute();
  };
  const result = useAsyncCallback(() => {
    if (executeParams.current) {
      return import_client_app_settings_v2.appSettingsObjectsClient.postAppSettingsOwnershipByObjectId({
        ...executeParams.current
      });
    }
  });
  return getCallbackResult4(result, execute);
}
function useAccessorPermissionsV2(params, options = { autoFetch: true, autoFetchOnUpdate: true }) {
  const onError = typeof options.onError === "function" ? options.onError : () => void 0;
  const result = useAsyncAbortable(
    (signal) => {
      return import_client_app_settings_v2.appSettingsObjectsClient.getAppSettingsPermissionByObjectIdAndAccessorId({
        ...params,
        abortSignal: signal
      });
    },
    [JSON.stringify(params)],
    {
      executeOnMount: options.autoFetch,
      executeOnUpdate: options.autoFetchOnUpdate,
      onError
    }
  );
  return getResult5(result);
}
function useUpdateAccessorPermissionsV2() {
  const executeParams = (0, import_react5.useRef)();
  const execute = async (params) => {
    executeParams.current = params;
    await result.execute();
  };
  const result = useAsyncCallback(() => {
    if (executeParams.current) {
      return import_client_app_settings_v2.appSettingsObjectsClient.putAppSettingsPermissionByObjectIdAndAccessorId({
        ...executeParams.current
      });
    }
  });
  return getCallbackResult4(result, execute);
}
function useDeleteAccessorPermissionsV2() {
  const executeParams = (0, import_react5.useRef)();
  const execute = async (params) => {
    executeParams.current = params;
    await result.execute();
  };
  const result = useAsyncCallback(() => {
    if (executeParams.current) {
      return import_client_app_settings_v2.appSettingsObjectsClient.deleteAppSettingsPermissionByObjectIdAndAccessorId({
        ...executeParams.current
      });
    }
  });
  return getCallbackResult4(result, execute);
}
function usePermissionsV2(params, options = { autoFetch: true, autoFetchOnUpdate: true }) {
  const onError = typeof options.onError === "function" ? options.onError : () => void 0;
  const result = useAsyncAbortable(
    (signal) => {
      return import_client_app_settings_v2.appSettingsObjectsClient.getAppSettingsPermissionsByObjectId({
        ...params,
        abortSignal: signal
      });
    },
    [JSON.stringify(params)],
    {
      executeOnMount: options.autoFetch,
      executeOnUpdate: options.autoFetchOnUpdate,
      onError
    }
  );
  return getResult5(result);
}
function useCreatePermissionsV2() {
  const executeParams = (0, import_react5.useRef)();
  const execute = async (params) => {
    executeParams.current = params;
    await result.execute();
  };
  const result = useAsyncCallback(() => {
    if (executeParams.current) {
      return import_client_app_settings_v2.appSettingsObjectsClient.postAppSettingsPermissionByObjectId({
        ...executeParams.current
      });
    }
  });
  return getCallbackResult4(result, execute);
}
function useAllUsersPermissionsV2(params, options = { autoFetch: true, autoFetchOnUpdate: true }) {
  const onError = typeof options.onError === "function" ? options.onError : () => void 0;
  const result = useAsyncAbortable(
    (signal) => {
      return import_client_app_settings_v2.appSettingsObjectsClient.getAppSettingsAllUsersPermissionByObjectId({
        ...params,
        abortSignal: signal
      });
    },
    [JSON.stringify(params)],
    {
      executeOnMount: options.autoFetch,
      executeOnUpdate: options.autoFetchOnUpdate,
      onError
    }
  );
  return getResult5(result);
}
function useUpdateAllUsersPermissionsV2() {
  const executeParams = (0, import_react5.useRef)();
  const execute = async (params) => {
    executeParams.current = params;
    await result.execute();
  };
  const result = useAsyncCallback(() => {
    if (executeParams.current) {
      return import_client_app_settings_v2.appSettingsObjectsClient.putAppSettingsAllUsersPermissionByObjectId({
        ...executeParams.current
      });
    }
  });
  return getCallbackResult4(result, execute);
}
function useDeleteAllUsersPermissionsV2() {
  const executeParams = (0, import_react5.useRef)();
  const execute = async (params) => {
    executeParams.current = params;
    await result.execute();
  };
  const result = useAsyncCallback(() => {
    if (executeParams.current) {
      return import_client_app_settings_v2.appSettingsObjectsClient.deleteAppSettingsAllUsersPermissionByObjectId({
        ...executeParams.current
      });
    }
  });
  return getCallbackResult4(result, execute);
}
function getCallbackResult4(result, execute) {
  const error = result.error;
  return {
    execute,
    data: result.result,
    error: error ? result.error : void 0,
    errorDetails: error?.body?.error,
    isError: error ? true : false,
    isSuccess: !error && !result.loading && result.result,
    isLoading: result.loading,
    status: result.status
  };
}
function getResult5(result) {
  const error = result.error;
  return {
    data: result.result,
    error: error ? result.error : void 0,
    errorDetails: error?.body?.error,
    isError: error ? true : false,
    isLoading: result.loading,
    isSuccess: !error && !result.loading && result.result,
    status: result.status,
    refetch: result.execute,
    cancel: result.cancel
  };
}

// packages/util/react-hooks/src/use-grail-fields/useGrailFields.ts
var import_react_query2 = require("@tanstack/react-query");

// packages/util/react-hooks/src/use-grail-fields/constants.ts
var import_react_query = require("@tanstack/react-query");
var DEFAULT_QUERY_CLIENT = new import_react_query.QueryClient({
  defaultOptions: {
    queries: {
      refetchOnMount: false,
      refetchOnWindowFocus: false,
      refetchOnReconnect: false,
      refetchInterval: false,
      staleTime: 6e4,
      retry: false
    }
  }
});

// packages/util/react-hooks/src/use-grail-fields/query-options.ts
var import_client_query3 = require("@dynatrace-sdk/client-query");

// packages/util/dql-utils/src/query/utils.ts
var import_client_query2 = require("@dynatrace-sdk/client-query");
var import_http_client = require("@dynatrace-sdk/http-client");
var delay2 = (ms, signal) => new Promise((resolve, reject) => {
  const listener = () => {
    clearTimeout(timer);
    reject(new DOMException("Delay aborted", "AbortDelayError"));
  };
  const timer = setTimeout(() => {
    signal?.removeEventListener("abort", listener);
    resolve();
  }, ms);
  signal?.addEventListener("abort", listener);
});
function isTooManyRequestsError(error) {
  return (0, import_client_query2.isErrorEnvelopeError)(error) && error.response.status === 429;
}
function getDelayFromError(error) {
  return error.body.error.retryAfterSeconds ?? (+error.response.headers["retryAfterSeconds"] || void 0);
}
var defaultQueryExecutionClient = new import_client_query2.QueryExecutionClient(import_http_client.httpClient, {
  transformDates: false
});
var clientCache = /* @__PURE__ */ new Map();
function getQueryExecutionClient(customHttpClient) {
  if (!customHttpClient)
    return defaultQueryExecutionClient;
  if (!clientCache.has(customHttpClient)) {
    clientCache.set(
      customHttpClient,
      new import_client_query2.QueryExecutionClient(customHttpClient, { transformDates: false })
    );
  }
  return clientCache.get(customHttpClient);
}
async function promiseStatus(promise) {
  const referenceObject = { name: "Victor" };
  const value = await Promise.race([promise, referenceObject]);
  return value === referenceObject ? "pending" : "fulfilled";
}
function isAbortDelayError(error) {
  return error instanceof DOMException && error.name === "AbortDelayError";
}

// packages/util/dql-utils/src/query/query.ts
var import_http_client2 = require("@dynatrace-sdk/http-client");
var IS_POLLING_DONE_INTERVAL = 500;
var IS_POLLING_DONE_TIMEOUT = 5e3;
async function runQuery(query, { onRequest, onExecute, onPoll, abortSignal, customHttpClient, pollRate } = {}) {
  const q = typeof query === "string" ? { query } : query;
  const queryExecutionClient3 = getQueryExecutionClient(customHttpClient);
  const executeCleanupController = new AbortController();
  const pollCleanupController = new AbortController();
  let executeQueryResponse = null;
  let timeout = 1;
  while (!executeQueryResponse?.result && !executeQueryResponse?.requestToken) {
    try {
      executeQueryResponse = await queryExecutionClient3.queryExecute({
        body: q,
        abortSignal,
        ...q
      });
    } catch (error) {
      if (isTooManyRequestsError(error)) {
        const delayTime = (getDelayFromError(error) ?? Math.min(timeout++ ** 2, 30)) * 1e3;
        try {
          await delay2(delayTime, abortSignal);
        } catch (delayError) {
          if (isAbortDelayError(delayError)) {
            return;
          }
          throw delayError;
        }
      } else if ((0, import_http_client2.isHttpClientAbortError)(error)) {
        return;
      } else {
        throw error;
      }
    }
  }
  if (onExecute)
    onExecute(executeQueryResponse);
  if (onRequest)
    onRequest(executeQueryResponse);
  if (!(executeQueryResponse?.state === "RUNNING" || executeQueryResponse?.state === "NOT_STARTED" || executeQueryResponse?.result === void 0)) {
    return executeQueryResponse.result;
  }
  async function cancelQuery2() {
    try {
      await queryExecutionClient3.queryCancel({
        requestToken: executeQueryResponse?.requestToken,
        ...q
      });
    } catch (error) {
      console.warn("Query cancellation failed:", error);
    }
  }
  abortSignal?.addEventListener("abort", cancelQuery2, { signal: executeCleanupController.signal });
  let pollQueryResponse = null;
  const abortPollController = new AbortController();
  while (!abortSignal?.aborted && executeQueryResponse.requestToken && (pollQueryResponse === null || pollQueryResponse.state === "RUNNING" || pollQueryResponse.state === "NOT_STARTED")) {
    try {
      await delay2(pollRate ?? 1e3, abortSignal);
      const pollPromise = queryExecutionClient3.queryPoll({
        requestToken: executeQueryResponse.requestToken,
        abortSignal: abortPollController.signal,
        ...q
      });
      if (!executeCleanupController.signal.aborted) {
        executeCleanupController.abort();
        abortSignal?.addEventListener(
          "abort",
          async () => {
            const isQueryCompleted = await new Promise((res) => {
              const interval = setInterval(async () => {
                if (pollQueryResponse?.state === "SUCCEEDED") {
                  clearInterval(interval);
                  clearTimeout(timeout2);
                  res(true);
                }
                const isPollingFinished = await promiseStatus(pollPromise) === "fulfilled";
                if (isPollingFinished && pollQueryResponse?.state !== "SUCCEEDED") {
                  clearInterval(interval);
                  clearTimeout(timeout2);
                  res(false);
                }
              }, IS_POLLING_DONE_INTERVAL);
              const timeout2 = setTimeout(() => {
                clearInterval(interval);
                abortPollController.abort();
                res(true);
              }, IS_POLLING_DONE_TIMEOUT);
            });
            if (!isQueryCompleted) {
              cancelQuery2();
            }
            pollCleanupController.abort();
          },
          { signal: pollCleanupController.signal }
        );
      }
      pollQueryResponse = await pollPromise;
      if (onPoll)
        onPoll(pollQueryResponse);
      if (onRequest)
        onRequest(pollQueryResponse);
    } catch (error) {
      if ((0, import_http_client2.isHttpClientAbortError)(error) || isAbortDelayError(error)) {
        return;
      }
      throw error;
    }
  }
  executeCleanupController.abort();
  pollCleanupController.abort();
  return pollQueryResponse?.result;
}

// packages/util/react-hooks/src/use-grail-fields/query-options.ts
function getGrailFieldsQueryOptions(autocompleteOptions, dssOptions) {
  const { query, ...autocompleteOverrides } = autocompleteOptions;
  const { dataObject, requestTimeoutMilliseconds, filterStatement, ...dssOverrides } = dssOptions;
  const dssQuery = filterStatement ? `internalLoadDss ${dataObject} | ${filterStatement}` : `internalLoadDss ${dataObject}`;
  return {
    autocomplete: {
      queryKey: ["useGrailFields", "autocomplete", query],
      queryFn: ({ signal }) => import_client_query3.queryAssistanceClient.queryAutocomplete({
        body: {
          query,
          cursorPosition: query.length
        },
        abortSignal: signal
      }),
      ...autocompleteOverrides
    },
    dss: {
      queryKey: ["useGrailFields", "dss", dssQuery, requestTimeoutMilliseconds],
      queryFn: ({ signal }) => runQuery(
        {
          query: dssQuery,
          requestTimeoutMilliseconds
        },
        { abortSignal: signal }
      ),
      ...dssOverrides
    }
  };
}

// packages/util/react-hooks/src/use-grail-fields/useGrailFields.ts
function useGrailFields(options) {
  const { autocompleteQueryOptions, dssQueryOptions, _queryClient, select } = options;
  const queryClient = _queryClient ?? DEFAULT_QUERY_CLIENT;
  const { autocomplete: autocompleteQuery, dss: dssQuery } = getGrailFieldsQueryOptions(
    autocompleteQueryOptions,
    dssQueryOptions
  );
  const [autocompleteResult, dssResult] = (0, import_react_query2.useQueries)(
    {
      queries: [autocompleteQuery, dssQuery]
    },
    queryClient
  );
  const rawData = {
    autocompleteData: autocompleteResult.data,
    dssData: dssResult.data
  };
  const isAutocompleteDone = autocompleteResult.isSuccess || autocompleteResult.fetchStatus === "idle" && !autocompleteResult.isError;
  const isDssDone = dssResult.isSuccess || dssResult.fetchStatus === "idle" && !dssResult.isError;
  const isSuccess = isAutocompleteDone && isDssDone;
  const selectedData = isSuccess ? select ? select(rawData) : rawData : void 0;
  return {
    data: selectedData,
    errors: {
      autocomplete: autocompleteResult.error,
      dss: dssResult.error
    },
    isError: autocompleteResult.isError || dssResult.isError,
    isLoading: autocompleteResult.isLoading || dssResult.isLoading,
    isSuccess,
    cancel: () => Promise.all([
      queryClient.cancelQueries({ queryKey: autocompleteQuery.queryKey }),
      queryClient.cancelQueries({ queryKey: dssQuery.queryKey })
    ]),
    refetch: () => Promise.all([autocompleteResult.refetch(), dssResult.refetch()])
  };
}

// packages/util/react-hooks/src/use-dql/useDql.ts
var import_react_query4 = require("@tanstack/react-query");
var import_react7 = require("react");

// packages/util/react-hooks/src/use-dql/query.ts
var import_client_query5 = require("@dynatrace-sdk/client-query");
var import_http_client3 = require("@dynatrace-sdk/http-client");

// packages/util/react-hooks/src/use-dql/query-client.ts
var import_react_query3 = require("@tanstack/react-query");
function getDefaultHookOptions() {
  return {
    pollRate: 1e3,
    runInBackground: false
  };
}
function getDefaultQueryClientConfig() {
  return {
    defaultOptions: {
      queries: {
        refetchOnMount: false,
        refetchOnWindowFocus: false,
        refetchOnReconnect: false,
        refetchInterval: false,
        staleTime: 6e4,
        retry: false
      }
    }
  };
}
function getQueryClientOptions(opts) {
  return Object.fromEntries(
    Object.entries({
      staleTime: opts.staleTime,
      refetchInterval: opts.refetchInterval,
      enabled: opts.enabled,
      placeholderData: opts.placeholderData,
      retryOnMount: opts.retryOnMount
    }).filter(([, value]) => value !== void 0)
  );
}
function createQueryKey(queryParams, customQueryKey) {
  const queryKey = customQueryKey ?? ["usedql", queryParams];
  return { queryKey, _getQueryKey: () => queryKey };
}
var defaultQueryClient = new import_react_query3.QueryClient(getDefaultQueryClientConfig());

// packages/util/react-hooks/src/use-dql/utils.ts
var import_client_query4 = require("@dynatrace-sdk/client-query");
var delay3 = (ms, value, signal) => new Promise((resolve, reject) => {
  const listener = () => {
    clearTimeout(timer);
    reject(value);
  };
  const timer = setTimeout(() => {
    signal?.removeEventListener("abort", listener);
    resolve(value);
  }, ms);
  signal?.addEventListener("abort", listener);
});
function isTooManyRequestsError2(error) {
  return (0, import_client_query4.isErrorEnvelopeError)(error) && error.response.status === 429;
}
function isString(val) {
  return typeof val === "string" || val instanceof String;
}
async function promiseStatus2(promise) {
  const referenceObject = { name: "Victor" };
  const value = await Promise.race([promise, referenceObject]);
  return value === referenceObject ? "pending" : "fulfilled";
}
function getDelayFromError2(error) {
  return error.body.error.retryAfterSeconds ?? (+error.response.headers["retryAfterSeconds"] || void 0);
}
function handleReturnValues({
  queryResult,
  state,
  progress,
  cancel,
  data,
  refetch,
  forceRefetch,
  enablePreview,
  _getQueryKey
}) {
  const baseReturnValue = {
    state,
    progress,
    cancel,
    data: queryResult.data,
    error: queryResult.error,
    isSuccess: queryResult.isSuccess,
    isError: queryResult.isError,
    isPending: queryResult.isPending,
    isLoading: queryResult.isLoading,
    isLoadingError: queryResult.isLoadingError,
    isRefetchError: queryResult.isRefetchError,
    isFetching: queryResult.isFetching,
    isFetched: queryResult.isFetched,
    isStale: queryResult.isStale,
    isPlaceholderData: queryResult.isPlaceholderData,
    refetch,
    forceRefetch,
    _getQueryKey
  };
  if (queryResult.isFetching) {
    const partialResultsAvailable = enablePreview && data && progress && progress < 100;
    return {
      ...baseReturnValue,
      state: state === "SUCCEEDED" ? "NOT_STARTED" : state,
      progress: progress === 100 ? 0 : progress,
      data: partialResultsAvailable ? data : queryResult.data,
      isPlaceholderData: partialResultsAvailable ? false : queryResult.isPlaceholderData
    };
  }
  return baseReturnValue;
}

// packages/util/react-hooks/src/use-dql/query.ts
var queryExecutionClient2 = new import_client_query5.QueryExecutionClient(import_http_client3.httpClient, { transformDates: false });
async function executeQuery(params, signal, controller) {
  try {
    return await queryExecutionClient2.queryExecute({
      body: params,
      abortSignal: signal,
      ...params
    });
  } catch (error) {
    controller?.abort();
    throw error;
  }
}
async function pollQuery(requestToken, params, controller) {
  try {
    return await queryExecutionClient2.queryPoll({
      requestToken,
      ...params
    });
  } catch (error) {
    controller?.abort();
    throw error;
  }
}
async function cancelQuery(requestToken, params, controller, onSettled) {
  controller?.abort();
  try {
    const response = await queryExecutionClient2.queryCancel({
      requestToken,
      ...params
    });
    onSettled?.(response ?? void 0);
    return response;
  } catch {
    onSettled?.();
  }
}
function getHookOptions(opts) {
  return {
    ...getDefaultHookOptions(),
    ...opts
  };
}
function normalizeQueryParams(query, ctxParams) {
  const queryString = isString(query) ? query : query.query;
  const params = isString(query) ? {} : query;
  return {
    ...ctxParams,
    ...params,
    query: queryString
  };
}

// packages/util/react-hooks/src/use-dql/context.ts
var import_react6 = require("react");
var DqlQueryParamsContext = (0, import_react6.createContext)({});
function DqlQueryParamsProvider(props) {
  const { children, ...params } = props;
  return (0, import_react6.createElement)(
    DqlQueryParamsContext.Provider,
    {
      value: params
    },
    children
  );
}
var _UseDqlQueryClientContext = (0, import_react6.createContext)(defaultQueryClient);
function _UseDqlQueryClientProvider({
  children,
  client
}) {
  return (0, import_react6.createElement)(_UseDqlQueryClientContext.Provider, { value: client }, children);
}

// packages/util/react-hooks/src/use-dql/useDql.ts
var IS_POLLING_DONE_INTERVAL2 = 500;
var IS_POLLING_DONE_TIMEOUT2 = 5e3;
function useDql(query, options = {}) {
  const queryParamsContext = (0, import_react7.useContext)(DqlQueryParamsContext);
  const queryClientContext = (0, import_react7.useContext)(_UseDqlQueryClientContext);
  const [progress, setProgress] = (0, import_react7.useState)(0);
  const [state, setState] = (0, import_react7.useState)("NOT_STARTED");
  const [data, setData] = (0, import_react7.useState)(void 0);
  const queryParams = (0, import_react7.useMemo)(
    () => normalizeQueryParams(query, queryParamsContext),
    [query, queryParamsContext]
  );
  const hookOpts = (0, import_react7.useMemo)(() => getHookOptions(options), [options]);
  const queryClient = hookOpts._queryClient ?? queryClientContext;
  const queryClientOpts = (0, import_react7.useMemo)(() => getQueryClientOptions(hookOpts), [hookOpts]);
  const { queryKey, _getQueryKey } = (0, import_react7.useMemo)(
    () => createQueryKey(queryParams, hookOpts._queryKey),
    [queryParams, hookOpts._queryKey]
  );
  const cancel = (0, import_react7.useCallback)(
    () => queryClient.cancelQueries({ queryKey }),
    [queryClient, queryKey]
  );
  const queryResult = (0, import_react_query4.useQuery)(
    {
      // eslint-disable-next-line @tanstack/query/exhaustive-deps
      queryKey,
      ...queryClientOpts,
      queryFn: async ({ signal }) => {
        setState("NOT_STARTED");
        setProgress(0);
        setData(void 0);
        const removeHandlersController = new AbortController();
        const cancelExecuteController = new AbortController();
        try {
          let executeQueryResponse = null;
          let timeout = 1;
          while (!executeQueryResponse?.result && !executeQueryResponse?.requestToken) {
            try {
              executeQueryResponse = await executeQuery(
                queryParams,
                signal,
                removeHandlersController
              );
            } catch (error) {
              if (isTooManyRequestsError2(error)) {
                const delayTime = (getDelayFromError2(error) ?? Math.min(timeout++ ** 2, 30)) * 1e3;
                await delay3(delayTime, void 0, signal);
              } else {
                throw error;
              }
            }
          }
          if (executeQueryResponse?.requestToken) {
            signal.addEventListener(
              "abort",
              () => {
                void cancelQuery(
                  executeQueryResponse.requestToken,
                  queryParams,
                  removeHandlersController,
                  (response) => {
                    setProgress(response?.progress ?? 0);
                    setState(response?.state ?? "CANCELLED");
                  }
                );
              },
              { signal: cancelExecuteController.signal }
            );
            addEventListener(
              "beforeunload",
              () => {
                if (queryClient.isFetching({ queryKey })) {
                  removeHandlersController.abort();
                  navigator.sendBeacon(
                    `/platform/storage/query/v1/query:cancel?request-token=${encodeURIComponent(
                      executeQueryResponse.requestToken
                    )}`
                  );
                }
              },
              { signal: removeHandlersController.signal }
            );
          }
          if (executeQueryResponse?.requestToken && !hookOpts.runInBackground) {
            addEventListener(
              "visibilitychange",
              () => {
                if (document.hidden && queryClient.isFetching({ queryKey })) {
                  void cancel().catch(() => void 0);
                }
              },
              { signal: removeHandlersController.signal }
            );
          }
          if (executeQueryResponse?.progress && progress !== 0) {
            setProgress(executeQueryResponse.progress);
          }
          if (executeQueryResponse?.state && state !== "NOT_STARTED") {
            setState(executeQueryResponse.state);
          }
          if (!(executeQueryResponse.state === "RUNNING" || executeQueryResponse.state === "NOT_STARTED" || executeQueryResponse.result === void 0)) {
            return executeQueryResponse.result;
          }
          let pollQueryResponse = null;
          while (!signal.aborted && executeQueryResponse.requestToken && (pollQueryResponse === null || pollQueryResponse.state === "RUNNING" || pollQueryResponse.state === "NOT_STARTED")) {
            await delay3(hookOpts.pollRate, void 0, signal);
            const pollPromise = pollQuery(
              executeQueryResponse.requestToken,
              queryParams,
              removeHandlersController
            );
            if (!cancelExecuteController.signal.aborted) {
              cancelExecuteController.abort();
              signal.addEventListener("abort", () => {
                void (async () => {
                  setProgress(0);
                  setState("CANCELLED");
                  const isQueryCompleted = await new Promise((res) => {
                    const interval = setInterval(() => {
                      if (pollQueryResponse?.state === "SUCCEEDED") {
                        clearInterval(interval);
                        clearTimeout(timeout2);
                        res(true);
                        return;
                      }
                      void promiseStatus2(pollPromise).then((status) => {
                        if (status === "fulfilled" && pollQueryResponse?.state !== "SUCCEEDED") {
                          clearInterval(interval);
                          clearTimeout(timeout2);
                          res(false);
                        }
                      }).catch(() => {
                        clearInterval(interval);
                        clearTimeout(timeout2);
                        res(false);
                      });
                    }, IS_POLLING_DONE_INTERVAL2);
                    const timeout2 = setTimeout(() => {
                      clearInterval(interval);
                      res(true);
                    }, IS_POLLING_DONE_TIMEOUT2);
                  });
                  if (!isQueryCompleted) {
                    await cancelQuery(
                      executeQueryResponse.requestToken,
                      queryParams,
                      removeHandlersController,
                      (response) => {
                        setProgress(response?.progress ?? 0);
                        setState(response?.state ?? "CANCELLED");
                      }
                    );
                  }
                })();
              });
            }
            pollQueryResponse = await pollPromise;
            const pollQueryResult = pollQueryResponse.result;
            if (queryParams.enablePreview && pollQueryResult && pollQueryResponse.state !== "SUCCEEDED") {
              setData(pollQueryResult);
            }
            if (!signal.aborted && pollQueryResponse?.progress) {
              setProgress(pollQueryResponse.progress);
            }
            if (!signal.aborted && pollQueryResponse?.state) {
              setState(pollQueryResponse.state);
            }
          }
          return pollQueryResponse?.result;
        } finally {
          removeHandlersController.abort();
          cancelExecuteController.abort();
        }
      }
    },
    queryClient
  );
  const {
    refetch: baseRefetch,
    isStale,
    isFetching,
    data: queryData,
    error: queryError,
    isSuccess,
    isError,
    isPending,
    isLoading,
    isLoadingError,
    isRefetchError,
    isFetched,
    isPlaceholderData
  } = queryResult;
  const { refetch, forceRefetch } = (0, import_react7.useMemo)(() => {
    const fns = {
      forceRefetch: async (options2) => {
        const refetchResult = await baseRefetch(options2);
        return handleReturnValues({
          queryResult: refetchResult,
          state,
          progress,
          data,
          cancel,
          refetch: fns.refetch,
          forceRefetch: fns.forceRefetch,
          enablePreview: queryParams.enablePreview,
          _getQueryKey
        });
      },
      refetch: async (options2) => {
        if ((isStale || !hookOpts.enabled) && !isFetching) {
          return fns.forceRefetch(options2);
        }
        return handleReturnValues({
          queryResult: {
            data: queryData,
            error: queryError,
            isSuccess,
            isError,
            isPending,
            isLoading,
            isLoadingError,
            isRefetchError,
            isFetching,
            isFetched,
            isStale,
            isPlaceholderData
          },
          state,
          progress,
          data,
          cancel,
          refetch: fns.refetch,
          forceRefetch: fns.forceRefetch,
          enablePreview: queryParams.enablePreview,
          _getQueryKey
        });
      }
    };
    return fns;
  }, [
    cancel,
    data,
    _getQueryKey,
    hookOpts.enabled,
    isFetching,
    isStale,
    baseRefetch,
    progress,
    queryParams.enablePreview,
    queryData,
    queryError,
    isSuccess,
    isError,
    isPending,
    isLoading,
    isLoadingError,
    isRefetchError,
    isFetched,
    isPlaceholderData,
    state
  ]);
  return handleReturnValues({
    queryResult,
    state,
    progress,
    data,
    cancel,
    refetch,
    forceRefetch,
    enablePreview: queryParams.enablePreview,
    _getQueryKey
  });
}

// packages/util/react-hooks/src/use-users/useUser.ts
var import_react_query6 = require("@tanstack/react-query");
var import_react9 = require("react");
var import_app_environment2 = require("@dynatrace-sdk/app-environment");

// packages/util/react-hooks/src/use-users/context.ts
var import_react8 = require("react");

// packages/util/react-hooks/src/use-users/query-client.ts
var import_react_query5 = require("@tanstack/react-query");
var STALE_TIME = 5 * 60 * 1e3;
var GC_TIME = 30 * 60 * 1e3;
function getDefaultUserQueryClientConfig() {
  return {
    defaultOptions: {
      queries: {
        refetchOnMount: false,
        refetchOnWindowFocus: false,
        refetchOnReconnect: false,
        refetchInterval: false,
        staleTime: STALE_TIME,
        gcTime: GC_TIME,
        retry: false
      }
    }
  };
}
function createUserQueryKey(environmentId, uuid) {
  return ["iam-user", environmentId, uuid];
}
function getUserQueryOptions(opts) {
  return Object.fromEntries(
    Object.entries({
      staleTime: opts.staleTime,
      enabled: opts.enabled
    }).filter(([, value]) => value !== void 0)
  );
}
var defaultUserQueryClient = new import_react_query5.QueryClient(getDefaultUserQueryClientConfig());

// packages/util/react-hooks/src/use-users/context.ts
var UseUsersConfigContext = (0, import_react8.createContext)({});
var UseUsersQueryClientContext = (0, import_react8.createContext)(defaultUserQueryClient);
function UseUsersProvider(props) {
  const { children, queryClient, ...config } = props;
  let content = (0, import_react8.createElement)(
    UseUsersConfigContext.Provider,
    { value: config },
    children
  );
  if (queryClient) {
    content = (0, import_react8.createElement)(UseUsersQueryClientContext.Provider, { value: queryClient }, content);
  }
  return content;
}

// packages/util/react-hooks/src/use-users/batch-loader.ts
var import_client_iam = require("@dynatrace-sdk/client-iam");
var import_app_environment = require("@dynatrace-sdk/app-environment");
var BATCH_WINDOW_MS = 50;
var MAX_CHUNK_SIZE = 100;
function chunk(array, size) {
  const chunks = [];
  for (let i = 0; i < array.length; i += size) {
    chunks.push(array.slice(i, i + size));
  }
  return chunks;
}
var UserBatchLoader = class {
  pending = [];
  scheduled = false;
  config;
  constructor(config) {
    this.config = config;
  }
  /**
   * Register a UUID for loading. Returns a promise that resolves with the user
   * data once the batch request completes, or `null` if the user was not found.
   *
   * If the UUID matches the currently logged-in user, resolves immediately
   * from the Dynatrace runtime without making an API call.
   */
  load(uuid, abortSignal) {
    const currentUser = (0, import_app_environment.getCurrentUserDetails)();
    if (currentUser.id === uuid) {
      return Promise.resolve({
        uid: currentUser.id,
        email: currentUser.email,
        name: currentUser.name
      });
    }
    return new Promise((resolve, reject) => {
      if (abortSignal?.aborted) {
        reject(abortSignal.reason ?? new DOMException("Aborted", "AbortError"));
        return;
      }
      const request = { uuid, resolve, reject };
      this.pending.push(request);
      abortSignal?.addEventListener(
        "abort",
        () => {
          reject(abortSignal.reason ?? new DOMException("Aborted", "AbortError"));
        },
        { once: true }
      );
      if (!this.scheduled) {
        this.scheduled = true;
        setTimeout(() => this.flush(), BATCH_WINDOW_MS);
      }
    });
  }
  async flush() {
    const batch = this.pending;
    this.pending = [];
    this.scheduled = false;
    if (batch.length === 0)
      return;
    const byUuid = /* @__PURE__ */ new Map();
    for (const req of batch) {
      const existing = byUuid.get(req.uuid);
      if (existing) {
        existing.push(req);
      } else {
        byUuid.set(req.uuid, [req]);
      }
    }
    const uniqueUuids = Array.from(byUuid.keys());
    const chunks = chunk(uniqueUuids, MAX_CHUNK_SIZE);
    const chunkPromises = chunks.map((uuidChunk) => this.fetchChunk(uuidChunk));
    const chunkResults = await Promise.allSettled(chunkPromises);
    for (let i = 0; i < chunkResults.length; i++) {
      const result = chunkResults[i];
      const uuidChunk = chunks[i];
      if (result.status === "fulfilled") {
        const usersByUid = /* @__PURE__ */ new Map();
        for (const user of result.value) {
          usersByUid.set(user.uid, user);
        }
        for (const uuid of uuidChunk) {
          const user = usersByUid.get(uuid) ?? null;
          const requests = byUuid.get(uuid);
          if (requests) {
            for (const req of requests) {
              req.resolve(user);
            }
          }
        }
      } else {
        const error = result.reason instanceof Error ? result.reason : new Error(String(result.reason));
        for (const uuid of uuidChunk) {
          const requests = byUuid.get(uuid);
          if (requests) {
            for (const req of requests) {
              req.reject(error);
            }
          }
        }
      }
    }
  }
  async fetchChunk(uuids) {
    const response = await import_client_iam.usersAndGroupsClient.getActiveUsersForOrganizationalLevelPost({
      body: uuids,
      levelType: this.config.levelType,
      levelId: this.config.levelId
    });
    return response.results ?? [];
  }
};

// packages/util/react-hooks/src/use-users/useUser.ts
var loaderCache = /* @__PURE__ */ new Map();
function getOrCreateLoader(levelType, levelId) {
  const key = `${levelType}:${levelId}`;
  let loader = loaderCache.get(key);
  if (!loader) {
    loader = new UserBatchLoader({ levelType, levelId });
    loaderCache.set(key, loader);
  }
  return loader;
}
function useUser(uuid, options = {}) {
  const ctxConfig = (0, import_react9.useContext)(UseUsersConfigContext);
  const queryClient = (0, import_react9.useContext)(UseUsersQueryClientContext);
  const levelType = options.levelType ?? ctxConfig.levelType ?? "environment";
  const levelId = options.levelId ?? ctxConfig.levelId ?? (0, import_app_environment2.getEnvironmentId)();
  const loader = (0, import_react9.useMemo)(() => getOrCreateLoader(levelType, levelId), [levelType, levelId]);
  const queryKey = (0, import_react9.useMemo)(
    () => uuid ? createUserQueryKey(levelId, uuid) : ["iam-user", levelId, ""],
    [levelId, uuid]
  );
  const queryClientOpts = (0, import_react9.useMemo)(() => getUserQueryOptions(options), [options]);
  const queryResult = (0, import_react_query6.useQuery)(
    {
      queryKey,
      ...queryClientOpts,
      enabled: !!uuid && options.enabled !== false,
      queryFn: ({ signal }) => {
        return loader.load(uuid, signal);
      }
    },
    queryClient
  );
  return {
    data: queryResult.data,
    error: queryResult.error,
    isPending: queryResult.isPending,
    isLoading: queryResult.isLoading,
    isSuccess: queryResult.isSuccess,
    isError: queryResult.isError,
    isFetching: queryResult.isFetching,
    refetch: queryResult.refetch
  };
}

// packages/util/react-hooks/src/use-users/useUsers.ts
var import_react_query7 = require("@tanstack/react-query");
var import_react10 = require("react");
var import_app_environment3 = require("@dynatrace-sdk/app-environment");
var loaderCache2 = /* @__PURE__ */ new Map();
function getOrCreateLoader2(levelType, levelId) {
  const key = `${levelType}:${levelId}`;
  let loader = loaderCache2.get(key);
  if (!loader) {
    loader = new UserBatchLoader({ levelType, levelId });
    loaderCache2.set(key, loader);
  }
  return loader;
}
function useUsers(uuids, options = {}) {
  const ctxConfig = (0, import_react10.useContext)(UseUsersConfigContext);
  const queryClient = (0, import_react10.useContext)(UseUsersQueryClientContext);
  const levelType = options.levelType ?? ctxConfig.levelType ?? "environment";
  const levelId = options.levelId ?? ctxConfig.levelId ?? (0, import_app_environment3.getEnvironmentId)();
  const loader = (0, import_react10.useMemo)(() => getOrCreateLoader2(levelType, levelId), [levelType, levelId]);
  const queryClientOpts = (0, import_react10.useMemo)(() => getUserQueryOptions(options), [options]);
  const queries = (0, import_react10.useMemo)(
    () => uuids.map((uuid) => ({
      queryKey: createUserQueryKey(levelId, uuid),
      ...queryClientOpts,
      enabled: options.enabled !== false,
      queryFn: ({ signal }) => loader.load(uuid, signal)
    })),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [JSON.stringify(uuids), levelId, queryClientOpts, options.enabled, loader]
  );
  const results = (0, import_react_query7.useQueries)({ queries }, queryClient);
  const resultData = results.map((r) => r.data);
  const resultErrors = results.map((r) => r.error);
  const refetchFns = results.map((r) => r.refetch);
  const data = (0, import_react10.useMemo)(() => {
    const map = /* @__PURE__ */ new Map();
    for (let i = 0; i < uuids.length; i++) {
      const d = resultData[i];
      if (d) {
        map.set(uuids[i], d);
      }
    }
    return map;
  }, [JSON.stringify(uuids), ...resultData]);
  const users = (0, import_react10.useMemo)(() => Array.from(data.values()), [data]);
  const errors = (0, import_react10.useMemo)(() => {
    const map = /* @__PURE__ */ new Map();
    for (let i = 0; i < uuids.length; i++) {
      const e = resultErrors[i];
      if (e) {
        map.set(uuids[i], e);
      }
    }
    return map;
  }, [JSON.stringify(uuids), ...resultErrors]);
  const isLoading = results.some((r) => r.isLoading);
  const isError = results.some((r) => r.isError);
  const isSuccess = results.length > 0 && results.every((r) => r.isSuccess);
  const isFetching = results.some((r) => r.isFetching);
  const refetch = (0, import_react10.useCallback)(() => {
    for (const fn of refetchFns) {
      fn();
    }
  }, [...refetchFns]);
  return {
    data,
    users,
    isLoading,
    isSuccess,
    isError,
    isFetching,
    errors,
    refetch
  };
}
