// @dynatrace/strato-components/esm/filters/segment-selector/contexts/SegmentsUI/segmentValidationHelper.js
import { jsx } from "react/jsx-runtime";
import { isArray } from "lodash-es";
import { FormattedMessage } from "react-intl";
import {
  FormFieldMessagesAction
} from "../../../../forms/forms-core/state/formfield-messages-state-reducer.js";
import { SEGMENTSELECTOR_MESSAGES } from "../../segmentMessages.js";
import {
  getSegment,
  segmentsClient
} from "../SegmentsCache/SegmentsCache.js";
async function getErrorKey(uiSegment) {
  if ("isPlaceholder" in uiSegment && uiSegment.isPlaceholder) {
    return null;
  }
  const cachedSegment = await segmentsClient.queryClient.fetchQuery({
    queryKey: ["getSegment", uiSegment.id],
    queryFn: () => getSegment(uiSegment.id).catch(() => null)
  });
  if (!cachedSegment) {
    return "unavailable";
  }
  const uiSegmentHasNoVariables = uiSegment.variables === void 0 || isArray(uiSegment.variables) && uiSegment.variables.length === 0;
  const cachedSegmentHasNoVariables = !cachedSegment.segment.variables;
  if (!cachedSegmentHasNoVariables && uiSegmentHasNoVariables) {
    return "needs-variables";
  }
  if (!uiSegmentHasNoVariables && cachedSegmentHasNoVariables) {
    return "needs-update";
  }
  return null;
}
async function areUiSegmentsValid(uiSegments) {
  if (segmentsClient.hasAccessRights === false || !uiSegments) {
    return void 0;
  }
  let hasError = false;
  await Promise.all(
    uiSegments?.map(async (uiSegment) => {
      const errorKey = await getErrorKey(uiSegment);
      if (errorKey) {
        hasError = true;
      }
      return errorKey;
    })
  );
  return !hasError;
}
async function validateSegments(segments, onChange, formFieldMessagesDispatch, formControlMessages, errorId, showError = false) {
  onChange();
  const isValid = await areUiSegmentsValid(segments);
  const hasFormControlError = formControlMessages.some(
    (message) => message.id === errorId
  );
  if (hasFormControlError && (segmentsClient.hasAccessRights === false && segments.length === 0 || isValid)) {
    formFieldMessagesDispatch?.({
      type: FormFieldMessagesAction.REMOVEMESSAGE,
      payload: {
        id: errorId,
        messageType: "formcontrol"
      }
    });
    return;
  }
  if (segmentsClient.hasAccessRights === false && segments.length > 0) {
    formFieldMessagesDispatch?.({
      type: FormFieldMessagesAction.ADDMESSAGE,
      payload: {
        id: errorId,
        variant: "error",
        validityStateKey: "customError",
        messageType: "formcontrol",
        showError: true,
        message: /* @__PURE__ */ jsx(
          FormattedMessage,
          {
            ...SEGMENTSELECTOR_MESSAGES.noReadAndWriteAccess
          }
        )
      }
    });
  }
  if (isValid === false && !hasFormControlError) {
    formFieldMessagesDispatch?.({
      type: FormFieldMessagesAction.ADDMESSAGE,
      payload: {
        id: errorId,
        variant: "error",
        validityStateKey: "customError",
        messageType: "formcontrol",
        showError,
        message: /* @__PURE__ */ jsx(FormattedMessage, { ...SEGMENTSELECTOR_MESSAGES.genericSegmentError })
      }
    });
  }
}
export {
  areUiSegmentsValid,
  getErrorKey,
  validateSegments
};
