// @dynatrace/strato-components/charts/timeseries/components/overlays/TimeseriesSelection/TimeseriesSelectionArea.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TimeseriesSelectionArea_exports = {};
__export(TimeseriesSelectionArea_exports, {
  TimeseriesSelectionArea: () => TimeseriesSelectionArea
});
module.exports = __toCommonJS(TimeseriesSelectionArea_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_TimeseriesSelectionAreaTooltip = require("./TimeseriesSelectionAreaTooltip.js");
var import_TimeseriesSelectionDatapoints = require("./TimeseriesSelectionDatapoints.js");
var import_useGridPlotSize = require("../../../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_explore_reducers = require("../../../../core/components/interactions-state-helpers/reducers/explore-reducers.js");
var import_DynamicSelectionArea = require("../../../../core/components/selection-area/DynamicSelectionArea.js");
var import_useSynchronizedSelection = require("../../../../core/components/shared-chart-interactions/hooks/useSynchronizedSelection.js");
var import_get_percentage_from_pixel = require("../../../../core/utils/get-percentage-from-pixel.js");
var import_get_selection_pixels = require("../../../../core/utils/get-selection-pixels.js");
var import_selectors = require("../../../hooks/store/selectors.js");
var import_useChartInteractions = require("../../../hooks/useChartInteractions.js");
var import_useTimeseriesChartId = require("../../../hooks/useTimeseriesChartId.js");
var import_useTimeseriesOverlayRef = require("../../../hooks/useTimeseriesOverlayRef.js");
var import_useTimeseriesSelectionAreaRef = require("../../../hooks/useTimeseriesSelectionAreaRef.js");
var import_useTimeseriesXFormatter = require("../../../hooks/useTimeseriesXFormatter.js");
var import_useXScale = require("../../../hooks/useXScale.js");
const TimeseriesSelectionArea = () => {
  const bottomFormatter = (0, import_useTimeseriesXFormatter.useTimeseriesXFormatter)();
  const selectionAreaRef = (0, import_useTimeseriesSelectionAreaRef.useTimeseriesSelectionAreaRef)();
  const { currentScale: xScale } = (0, import_useXScale.useXScales)();
  const chartRef = (0, import_useTimeseriesOverlayRef.useTimeseriesOverlayRef)();
  const setState = (0, import_selectors.useSetState)();
  const selectState = (0, import_selectors.useSelectState)();
  const { zoomX } = (0, import_useChartInteractions.useChartInteractions)();
  const { width } = (0, import_useGridPlotSize.useGridPlotSize)();
  const interactionMode = (0, import_selectors.useInteractionModeState)();
  const { updateSelectionState } = (0, import_useSynchronizedSelection.useSynchronizedSelection)();
  const chartId = (0, import_useTimeseriesChartId.useTimeseriesChartId)();
  const hasBottomMagnifiers = !(0, import_lodash_es.isUndefined)(bottomFormatter);
  if ((0, import_lodash_es.isUndefined)(selectState) || !chartRef.current) {
    return null;
  }
  const { start, end, selectionPercentage, finished, accessible, variant } = selectState;
  const handleWrapperKeyDown = (event) => {
    const canZoom = zoomX?.enabled;
    if (event.key === "Enter" && canZoom) {
      chartRef.current?.focus();
    }
  };
  const handleSelectionChange = ({
    selection,
    accessible: accessible2
  }) => {
    setState(
      (0, import_explore_reducers.reduceExploreSelect)({
        start: selection[0],
        end: selection[1],
        finished: true,
        accessible: accessible2,
        variant,
        selectionPercentage: (0, import_get_percentage_from_pixel.getPercentageFromPixel)(selection, width)
      })
    );
    if (interactionMode === "explore") {
      updateSelectionState?.({
        selection,
        chartId,
        scale: xScale
      });
    }
  };
  const selectionPosition = finished ? (0, import_get_selection_pixels.getSelectionPixels)(selectionPercentage, width, [start, end]) : [start, end];
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    variant === "select" ? /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TimeseriesSelectionDatapoints.TimeseriesSelectionDatapoints, {}),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TimeseriesSelectionAreaTooltip.TimeseriesSelectionAreaTooltip, {})
    ] }) : null,
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_DynamicSelectionArea.DynamicSelectionArea,
      {
        chartRef,
        hasMagnifiers: hasBottomMagnifiers,
        selection: selectionPosition,
        variant,
        focusable: accessible,
        interactable: finished,
        bottomFormatter: (value) => bottomFormatter(xScale.invert(value)),
        onKeyDown: handleWrapperKeyDown,
        onSelectionChange: handleSelectionChange,
        selectionAreaRef
      }
    )
  ] });
};
