// @dynatrace/strato-components/charts/timeseries/components/overlays/TimeseriesSelection/TimeseriesSelectionAreaTooltip.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TimeseriesSelectionAreaTooltip_exports = {};
__export(TimeseriesSelectionAreaTooltip_exports, {
  TimeseriesSelectionAreaTooltip: () => TimeseriesSelectionAreaTooltip
});
module.exports = __toCommonJS(TimeseriesSelectionAreaTooltip_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_convert_to_timeseries_array = require("./utils/convert-to-timeseries-array.js");
var import_get_selected_datapoints = require("./utils/get-selected-datapoints.js");
var import_is_datapoint_within_selection = require("./utils/is-datapoint-within-selection.js");
var import_useGridPlotSize = require("../../../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_get_custom_context_menu_items = require("../../../../core/components/interactions-overlay/utils/get-custom-context-menu-items.js");
var import_OverlayTooltip = require("../../../../core/components/overlay-tooltip/OverlayTooltip.js");
var import_SelectionTooltipWithAsyncActions = require("../../../../core/components/selection-area/SelectionTooltipWithAsyncActions.js");
var import_get_selection_tooltip_position = require("../../../../core/components/selection-area/tooltip/get-selection-tooltip-position.js");
var import_usePointerAndScrollResponsiveChartBounds = require("../../../../core/hooks/usePointerAndScrollResponsiveChartBounds.js");
var import_selectors = require("../../../hooks/store/selectors.js");
var import_state_setters = require("../../../hooks/store/state-setters.js");
var import_useChartInteractions = require("../../../hooks/useChartInteractions.js");
var import_useTimeseriesFilteredData = require("../../../hooks/useTimeseriesFilteredData.js");
var import_useTimeseriesOverlayRef = require("../../../hooks/useTimeseriesOverlayRef.js");
var import_useTimeseriesSelect = require("../../../hooks/useTimeseriesSelect.js");
var import_useXScale = require("../../../hooks/useXScale.js");
var import_tooltips_text = require("../../../messages/tooltips-text.js");
var import_get_datapoints_message = require("../../../utils/get-datapoints-message.js");
var import_timeseries = require("../../../utils/timeseries.js");
const TimeseriesSelectionAreaTooltip = () => {
  const selectState = (0, import_selectors.useSelectState)();
  const { zoomX } = (0, import_useChartInteractions.useChartInteractions)();
  const timeseriesSelect = (0, import_useTimeseriesSelect.useTimeseriesSelect)();
  const { currentScale: xScale } = (0, import_useXScale.useXScales)();
  const zoomToSelection = (0, import_state_setters.useZoomToSelection)();
  const { data, allDatapoints, datapointsMap } = (0, import_useTimeseriesFilteredData.useTimeseriesFilteredData)();
  const { width } = (0, import_useGridPlotSize.useGridPlotSize)();
  const intl = (0, import_react_intl.useIntl)();
  const chartRef = (0, import_useTimeseriesOverlayRef.useTimeseriesOverlayRef)();
  const chartBounds = (0, import_usePointerAndScrollResponsiveChartBounds.usePointerAndScrollResponsiveChartBounds)(
    chartRef,
    true
  )();
  if (!selectState || !chartBounds) {
    return null;
  }
  const selectionStart = selectState.selectionPercentage ? width * selectState.selectionPercentage[0] / 100 : selectState.start.relative.x;
  const selectionEnd = selectState.selectionPercentage ? width * selectState.selectionPercentage[1] / 100 : selectState.end.relative.x;
  const range = [
    xScale.invert(selectionStart),
    xScale.invert(selectionEnd)
  ];
  const selectedDatapoints = (0, import_get_selected_datapoints.getSelectedDatapoints)(
    allDatapoints,
    datapointsMap,
    range
  );
  let totalSeries = 0;
  let totalDatapoints = 0;
  const [from, to] = range;
  for (const series of data) {
    let isSeriesPresent = false;
    for (const datapoint of series.datapoints) {
      const datapointDate = new Date(
        (0, import_timeseries.getDatapointTime)(datapoint.start, datapoint.end)
      );
      if ((0, import_is_datapoint_within_selection.isDatapointWithinSelection)(datapointDate, from, to)) {
        totalDatapoints++;
        isSeriesPresent = true;
      }
    }
    if (isSeriesPresent) {
      totalSeries++;
    }
  }
  const datapointsMessage = (0, import_get_datapoints_message.getDatapointsMessage)(totalDatapoints, data.length);
  const tooltipPosition = (0, import_get_selection_tooltip_position.getSelectionTooltipPosition)(selectState, chartBounds);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_SelectionTooltipWithAsyncActions.SelectionTooltipWithAsyncActions,
    {
      actions: () => {
        const actions = timeseriesSelect?.actions;
        return actions && (async () => actions(
          (0, import_convert_to_timeseries_array.convertToTimeseriesArray)(selectedDatapoints),
          range
        ))().then(import_get_custom_context_menu_items.getCustomContextMenuItems);
      },
      selectFinished: selectState.finished,
      header: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_OverlayTooltip.OverlayTooltip.Header,
        {
          header1: [from, to],
          showRangeDuration1: true,
          subheader: intl.formatMessage(datapointsMessage, {
            count: totalDatapoints,
            seriesCount: totalSeries
          })
        }
      ),
      position: tooltipPosition,
      zoomEnabled: !!zoomX?.enabled,
      range: selectState.selectionPercentage || [0, 0],
      children: zoomX?.enabled ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_OverlayTooltip.OverlayTooltip.Action,
        {
          label: intl.formatMessage(import_tooltips_text.TIMESERIES_MESSAGES.zoomToSelection),
          icon: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ZoomToSelectionIcon, {}),
          onSelect: zoomToSelection,
          trailingIcon: "\u23CE"
        }
      ) : null
    }
  );
};
