// @dynatrace/strato-components/esm/charts/meter-bar/shared/constants.js
import Colors from "@dynatrace/strato-design-tokens/colors";
const INITIAL_STATE = Object.freeze({
  position: {
    absolute: { x: 0, y: 0 },
    relative: { x: 0, y: 0 }
  },
  pinned: false,
  visible: false
});
const METER_BAR_SIZE = Object.freeze({
  Size16: "size16",
  Size24: "size24",
  Size8: "size8",
  Auto: "auto"
});
const MIN_DEFAULT_VALUE = 0;
const MAX_DEFAULT_VALUE = 100;
const SIZE_DEFAULT = METER_BAR_SIZE.Size16;
const DEFAULT_FORMATTER = (v) => v.toString();
const DEFAULT_GAP_WIDTH = 1;
const DEFAULT_WIDTH = "100%";
const DEFAULT_HEIGHT = "100%";
const SMALL_CHART_SIZE = 8;
const MEDIUM_CHART_SIZE = 16;
const LARGE_CHART_SIZE = 24;
const METER_BAR_FONT_COLOR = Colors.Text.Neutral.Default;
const METER_BAR_MIN_MAX_FONT_COLOR = Colors.Text.Neutral.Subdued;
export {
  DEFAULT_FORMATTER,
  DEFAULT_GAP_WIDTH,
  DEFAULT_HEIGHT,
  DEFAULT_WIDTH,
  INITIAL_STATE,
  LARGE_CHART_SIZE,
  MAX_DEFAULT_VALUE,
  MEDIUM_CHART_SIZE,
  METER_BAR_FONT_COLOR,
  METER_BAR_MIN_MAX_FONT_COLOR,
  METER_BAR_SIZE,
  MIN_DEFAULT_VALUE,
  SIZE_DEFAULT,
  SMALL_CHART_SIZE
};
