// @dynatrace/strato-components/esm/charts/core/components/annotations/components/canvas/AnnotationsCanvasLayout.js
import { jsx } from "react/jsx-runtime";
import { canvasContainer } from "./AnnotationsCanvasLayout.sty.js";
import { DisplayConfigContext } from "./context/display-config.context.js";
import { AnnotationsChartDisplay } from "./displays/AnnotationsChartDisplay.js";
import { Canvas } from "../../../canvas/Canvas.js";
import { CanvasContext } from "../../../canvas/CanvasContext.js";
import { useAnnotationsIconImages } from "../../hooks/useAnnotationsIconImages.js";
const AnnotationsCanvasLayout = ({
  height,
  width
}) => {
  const icons = useAnnotationsIconImages();
  const annotationsArea = { width, height, icons };
  return /* @__PURE__ */ jsx(DisplayConfigContext.Provider, { value: { annotationsArea }, children: /* @__PURE__ */ jsx(
    "div",
    {
      className: canvasContainer,
      style: {
        width,
        height,
        overflowY: "hidden"
      },
      children: /* @__PURE__ */ jsx(
        CanvasContext.Provider,
        {
          value: {
            shapeRefs: /* @__PURE__ */ new Map()
          },
          children: /* @__PURE__ */ jsx(Canvas, { width, height, children: /* @__PURE__ */ jsx(AnnotationsChartDisplay, {}) })
        }
      )
    }
  ) });
};
AnnotationsCanvasLayout["displayName"] = "AnnotationsCanvasLayout";
export {
  AnnotationsCanvasLayout
};
