// @dynatrace/strato-components/esm/charts/pie/components/PieChartPlot/PieChartPlot.js
import { jsx, jsxs } from "react/jsx-runtime";
import { Group } from "@visx/group";
import { Pie } from "@visx/shape";
import { arc } from "d3-shape";
import { forwardRef, Fragment, useMemo } from "react";
import { container } from "./PieChartPlot.sty.js";
import { useGraphSize } from "../../../core/components/chart-layout/hooks/useGraphSize.js";
import { useTooltipReducer } from "../../../core/components/overlay-tooltip/useTooltipReducer.js";
import { CHART_VARIANT, PAD_ANGLE } from "../../constants.js";
import {
  useFormatter,
  useTruncationMode
} from "../../context/FormatterContext.js";
import { useLabels } from "../../context/LabelsContext.js";
import {
  useHighlightedIdState,
  useVisibleSlicesState
} from "../../hooks/store/selectors.js";
import { buildChartSizes } from "../../utils/build-chart-sizes.js";
import { buildLabel } from "../../utils/build-label.js";
import { collision, isInBounds } from "../../utils/label-collisions.js";
import { InnerWrapper } from "../InnerWrapper/InnerWrapper.js";
import { PieChartLabel } from "../PieChartLabel.js";
import { PieChartSlice } from "../PieChartSlice.js";
const getValue = (pieData) => pieData.relativeValue;
const PieChartPlot = forwardRef(
  (props, forwardedRef) => {
    const { variant } = props;
    const dispatchReducer = useTooltipReducer();
    const visibleSlices = useVisibleSlicesState();
    const highlightedId = useHighlightedIdState();
    const { height, width } = useGraphSize();
    const { hidden: hiddenLabels, valueType } = useLabels();
    const formatter = useFormatter();
    const truncationMode = useTruncationMode();
    const { topCenter, radius, leftCenter, innerRadius } = buildChartSizes(
      height,
      width,
      variant,
      visibleSlices,
      !!hiddenLabels
    );
    const visibleIds = visibleSlices.map((slice) => slice.id);
    const donutOverlay = useMemo(
      () => arc().innerRadius(innerRadius).outerRadius(radius).startAngle(0).endAngle(360)({}),
      [innerRadius, radius]
    );
    return /* @__PURE__ */ jsxs("div", { className: container, ref: forwardedRef, children: [
      variant === CHART_VARIANT.Donut && /* @__PURE__ */ jsx(
        InnerWrapper,
        {
          width,
          height,
          innerRadius,
          visibleIds
        }
      ),
      /* @__PURE__ */ jsx("svg", { width, height, style: { overflow: "visible" }, children: /* @__PURE__ */ jsxs(
        Group,
        {
          top: topCenter,
          left: leftCenter,
          onMouseLeave: dispatchReducer({ type: "SET_OUT_BOUNDS" }),
          children: [
            variant === CHART_VARIANT.Pie ? /* @__PURE__ */ jsx(
              "circle",
              {
                cx: "0",
                cy: "0",
                r: radius > 0 ? radius : 0,
                fill: "transparent"
              }
            ) : /* @__PURE__ */ jsx("path", { d: donutOverlay, fill: "transparent" }),
            /* @__PURE__ */ jsx(
              Pie,
              {
                data: visibleSlices,
                pieValue: getValue,
                outerRadius: radius,
                innerRadius,
                padAngle: PAD_ANGLE,
                pieSort: null,
                children: (pie) => {
                  let previousLabel = void 0;
                  return pie.arcs.map((arc2) => {
                    const label = buildLabel(
                      arc2,
                      pie.path,
                      valueType,
                      formatter,
                      truncationMode,
                      leftCenter
                    );
                    const isHighlighted = !highlightedId || highlightedId === arc2.data.id;
                    const showLabel = !hiddenLabels && isInBounds(label, leftCenter, topCenter) && (!previousLabel || !collision(label, previousLabel)) && arc2.data.id !== "thinner-slices";
                    if (showLabel) {
                      previousLabel = label;
                    }
                    return /* @__PURE__ */ jsxs(Fragment, { children: [
                      /* @__PURE__ */ jsx(
                        PieChartSlice,
                        {
                          sliceId: `slice-${arc2.data.id}`,
                          arc: arc2,
                          data: arc2.data,
                          path: pie.path,
                          isHighlighted
                        }
                      ),
                      showLabel && /* @__PURE__ */ jsx(
                        PieChartLabel,
                        {
                          x: label.textX,
                          y: label.textY,
                          text: label.text,
                          isHighlighted
                        }
                      )
                    ] }, `slice-${arc2.data.id}`);
                  });
                }
              }
            )
          ]
        }
      ) })
    ] });
  }
);
export {
  PieChartPlot
};
