// @dynatrace/strato-components/core/utils/create-dom-rect.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var create_dom_rect_exports = {};
__export(create_dom_rect_exports, {
  createDOMRect: () => createDOMRect
});
module.exports = __toCommonJS(create_dom_rect_exports);
function createDOMRect(position, size) {
  const { x = 0, y = 0 } = position;
  return () => {
    const values = {
      bottom: y,
      height: size ? size.height : 0,
      width: size ? size.width : 0,
      top: y,
      left: x,
      right: x,
      x,
      y
    };
    return {
      ...values,
      toJSON: () => JSON.parse(JSON.stringify(values))
    };
  };
}
