// @dynatrace/strato-components/core/utils/default-formatters.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var default_formatters_exports = {};
__export(default_formatters_exports, {
  DATETIME_DEFAULT_FORMAT_OPTIONS: () => DATETIME_DEFAULT_FORMAT_OPTIONS,
  formatCurrencyDefault: () => formatCurrencyDefault,
  formatDatetimeDefault: () => formatDatetimeDefault
});
module.exports = __toCommonJS(default_formatters_exports);
var import_units = require("@dynatrace-sdk/units");
const DATETIME_DEFAULT_FORMAT_OPTIONS = {
  year: "numeric",
  month: "numeric",
  day: "numeric",
  hour: "numeric",
  minute: "numeric",
  second: "numeric"
};
function formatDatetimeDefault(value) {
  const valueAsDate = value instanceof Date ? value : new Date(value);
  if (Number.isNaN(valueAsDate.getTime())) {
    return "";
  }
  return (0, import_units.formatDate)(valueAsDate, DATETIME_DEFAULT_FORMAT_OPTIONS);
}
function formatCurrencyDefault(value, formatOptions = {}) {
  return (0, import_units.format)(value, {
    abbreviate: false,
    minimumFractionDigits: 2,
    maximumFractionDigits: 2,
    ...formatOptions
  });
}
