// @dynatrace/strato-components/core/utils/framer-motion-easings.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var framer_motion_easings_exports = {};
__export(framer_motion_easings_exports, {
  closeEasing: () => closeEasing,
  expandCollapseEasing: () => expandCollapseEasing,
  fadeInEasing: () => fadeInEasing,
  fadeOutEasing: () => fadeOutEasing,
  hoverEasing: () => hoverEasing,
  moveEasing: () => moveEasing,
  openEasing: () => openEasing,
  rescaleEasing: () => rescaleEasing,
  slideInEasing: () => slideInEasing,
  slideOutEasing: () => slideOutEasing,
  staggeredMotionEasing: () => staggeredMotionEasing
});
module.exports = __toCommonJS(framer_motion_easings_exports);
var import_framer_motion = require("framer-motion");
var import_animations_js = __toESM(require("@dynatrace/strato-design-tokens/animations-js"));
const openEasing = (0, import_framer_motion.cubicBezier)(
  import_animations_js.default.Open.Easing[0],
  import_animations_js.default.Open.Easing[1],
  import_animations_js.default.Open.Easing[2],
  import_animations_js.default.Open.Easing[3]
);
const closeEasing = (0, import_framer_motion.cubicBezier)(
  import_animations_js.default.Close.Easing[0],
  import_animations_js.default.Close.Easing[1],
  import_animations_js.default.Close.Easing[2],
  import_animations_js.default.Close.Easing[3]
);
const slideInEasing = (0, import_framer_motion.cubicBezier)(
  import_animations_js.default.SlideIn.Easing[0],
  import_animations_js.default.SlideIn.Easing[1],
  import_animations_js.default.SlideIn.Easing[2],
  import_animations_js.default.SlideIn.Easing[3]
);
const slideOutEasing = (0, import_framer_motion.cubicBezier)(
  import_animations_js.default.SlideOut.Easing[0],
  import_animations_js.default.SlideOut.Easing[1],
  import_animations_js.default.SlideOut.Easing[2],
  import_animations_js.default.SlideOut.Easing[3]
);
const fadeInEasing = (0, import_framer_motion.cubicBezier)(
  import_animations_js.default.FadeIn.Easing[0],
  import_animations_js.default.FadeIn.Easing[1],
  import_animations_js.default.FadeIn.Easing[2],
  import_animations_js.default.FadeIn.Easing[3]
);
const fadeOutEasing = (0, import_framer_motion.cubicBezier)(
  import_animations_js.default.FadeOut.Easing[0],
  import_animations_js.default.FadeOut.Easing[1],
  import_animations_js.default.FadeOut.Easing[2],
  import_animations_js.default.FadeOut.Easing[3]
);
const hoverEasing = (0, import_framer_motion.cubicBezier)(
  import_animations_js.default.Hover.Easing[0],
  import_animations_js.default.Hover.Easing[1],
  import_animations_js.default.Hover.Easing[2],
  import_animations_js.default.Hover.Easing[3]
);
const staggeredMotionEasing = (0, import_framer_motion.cubicBezier)(
  import_animations_js.default.StaggeredMotion.Easing[0],
  import_animations_js.default.StaggeredMotion.Easing[1],
  import_animations_js.default.StaggeredMotion.Easing[2],
  import_animations_js.default.StaggeredMotion.Easing[3]
);
const expandCollapseEasing = (0, import_framer_motion.cubicBezier)(
  import_animations_js.default.ExpandCollapse.Easing[0],
  import_animations_js.default.ExpandCollapse.Easing[1],
  import_animations_js.default.ExpandCollapse.Easing[2],
  import_animations_js.default.ExpandCollapse.Easing[3]
);
const moveEasing = (0, import_framer_motion.cubicBezier)(
  import_animations_js.default.Move.Easing[0],
  import_animations_js.default.Move.Easing[1],
  import_animations_js.default.Move.Easing[2],
  import_animations_js.default.Move.Easing[3]
);
const rescaleEasing = (0, import_framer_motion.cubicBezier)(
  import_animations_js.default.Rescale.Easing[0],
  import_animations_js.default.Rescale.Easing[1],
  import_animations_js.default.Rescale.Easing[2],
  import_animations_js.default.Rescale.Easing[3]
);
