// @dynatrace/strato-components/core/utils/modality.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var modality_exports = {};
__export(modality_exports, {
  initModalityDetection: () => initModalityDetection,
  modalityStore: () => modalityStore
});
module.exports = __toCommonJS(modality_exports);
var import_store = require("../store/store.js");
const modalityStore = (0, import_store.createStore)("unknown");
const ignoreMap = [
  "Alt",
  "AltGraph",
  "CapsLock",
  "Control",
  "FnLock",
  "Fn",
  "Hyper",
  "Meta",
  // Command (Mac), in some browsers also Windows-key
  "NumLock",
  "OS",
  // Windows-key
  "ScrollLock",
  "Super",
  // Virtual keyboard key
  "Symbol",
  "SymbolLock"
];
const eventTypeMap = {
  keydown: "keyboard",
  keyup: "keyboard",
  mousedown: "mouse",
  mousemove: "mouse",
  MSPointerDown: "pointer",
  MSPointerMove: "pointer",
  pointerdown: "pointer",
  pointermove: "pointer",
  touchstart: "touch",
  touchend: "touch",
  wheel: "wheel"
};
const navigationalKeys = [
  "Tab",
  "Esc",
  "Escape",
  " ",
  "ArrowDown",
  "Down",
  "ArrowUp",
  "Up",
  "ArrowLeft",
  "Left",
  "ArrowRight",
  "Right",
  "End",
  "Home",
  "PageDown",
  "PageUp"
];
function handleKeyboardInput(event) {
  if (ignoreMap.includes(event.key)) {
    modalityStore.setState("pointer");
    return;
  }
  if (!event.target || navigationalKeys.includes(event.key)) {
    modalityStore.setState("keyboard");
  }
}
function handlePointerInput(event) {
  const modality = eventTypeMap[event.type];
  switch (modality) {
    case "touch":
    case "pen":
      modalityStore.setState("touch");
      break;
    case "mouse":
    case "pointer":
    case "wheel":
      modalityStore.setState("pointer");
      break;
  }
}
let isInitialized = false;
function initModalityDetection() {
  if (isInitialized) {
    return;
  }
  window.addEventListener("pointerdown", handlePointerInput);
  window.addEventListener("keydown", handleKeyboardInput, { capture: true });
  window.addEventListener("wheel", handlePointerInput);
  isInitialized = true;
}
