// @dynatrace/strato-components/core/utils/prism-theme.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var prism_theme_exports = {};
__export(prism_theme_exports, {
  DTPrismTheme: () => DTPrismTheme
});
module.exports = __toCommonJS(prism_theme_exports);
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
const DTPrismBaseTheme = {
  plain: {
    color: import_colors.default.Text.Neutral.Default,
    backgroundColor: "transparent"
  },
  styles: [
    {
      types: ["comment"],
      style: {
        color: import_colors.default.Syntax.Comment,
        fontStyle: "italic"
      }
    },
    {
      types: ["string", "builtin", "char", "constant", "url", "attr-value"],
      style: {
        color: import_colors.default.Syntax.String
      }
    },
    {
      types: [
        "variable",
        "attr-name",
        "function-variable",
        "maybe-class-name",
        "function",
        "property-access"
      ],
      style: {
        color: import_colors.default.Syntax.Variable
      }
    },
    {
      types: ["number"],
      style: {
        color: import_colors.default.Syntax.Number
      }
    },
    {
      types: ["punctuation"],
      style: {
        color: void 0
      }
    },
    {
      types: ["selector", "doctype", "class-name", "tag"],
      style: {
        color: import_colors.default.Syntax.Classname
      }
    },
    {
      types: ["keyword", "namespace"],
      style: {
        color: import_colors.default.Syntax.Operators
      }
    },
    {
      types: ["boolean"],
      style: {
        color: import_colors.default.Syntax.Boolean
      }
    }
  ]
};
const DTPrismTheme = {
  Default: {
    plain: DTPrismBaseTheme.plain,
    styles: [
      ...DTPrismBaseTheme.styles,
      {
        types: ["property"],
        style: {
          color: import_colors.default.Syntax.Variable
        }
      }
    ]
  },
  Log: {
    plain: DTPrismBaseTheme.plain,
    styles: [
      ...DTPrismBaseTheme.styles,
      {
        types: ["property"],
        style: {
          color: import_colors.default.Text.Neutral.Default
        }
      }
    ]
  }
};
