// @dynatrace/strato-components/forms/calendar/types/calendarConfiguration.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var calendarConfiguration_exports = {};
__export(calendarConfiguration_exports, {
  COLUMNS_GAP: () => COLUMNS_GAP,
  NUMBER_OF_MONTHS: () => NUMBER_OF_MONTHS,
  NUMBER_OF_MONTHS_PER_ROW: () => NUMBER_OF_MONTHS_PER_ROW,
  NUMBER_OF_MONTH_ROWS: () => NUMBER_OF_MONTH_ROWS,
  NUMBER_OF_WEEKDAYS: () => NUMBER_OF_WEEKDAYS,
  NUMBER_OF_WEEK_ROWS: () => NUMBER_OF_WEEK_ROWS,
  NUMBER_OF_YEARS: () => NUMBER_OF_YEARS,
  NUMBER_OF_YEARS_PER_ROW: () => NUMBER_OF_YEARS_PER_ROW,
  NUMBER_OF_YEAR_ROWS: () => NUMBER_OF_YEAR_ROWS,
  YEARS_RANGE_START_OFFSET: () => YEARS_RANGE_START_OFFSET
});
module.exports = __toCommonJS(calendarConfiguration_exports);
var import_spacings = __toESM(require("@dynatrace/strato-design-tokens/spacings"));
const NUMBER_OF_WEEKDAYS = 7;
const NUMBER_OF_MONTHS_PER_ROW = 3;
const NUMBER_OF_YEARS_PER_ROW = 4;
const NUMBER_OF_MONTHS = 12;
const NUMBER_OF_WEEK_ROWS = 6;
const NUMBER_OF_MONTH_ROWS = 4;
const NUMBER_OF_YEAR_ROWS = 6;
const COLUMNS_GAP = import_spacings.default.Size4;
const NUMBER_OF_YEARS = NUMBER_OF_YEAR_ROWS * 4;
const YEARS_RANGE_START_OFFSET = -5;
