// @dynatrace/strato-components/charts/categorical-bar/types/defaults.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var defaults_exports = {};
__export(defaults_exports, {
  DEFAULT_CHART_HEIGHT: () => DEFAULT_CHART_HEIGHT,
  DEFAULT_LEGEND: () => DEFAULT_LEGEND,
  defaultConfig: () => defaultConfig
});
module.exports = __toCommonJS(defaults_exports);
var import_constants = require("../../core/constants.js");
const defaultConfig = {
  layout: "vertical",
  valueRepresentation: "absolute",
  groupMode: "stacked",
  colorPalette: "categorical",
  tooltip: {
    variant: "single"
  },
  legend: {
    position: "auto",
    hidden: false,
    ratio: import_constants.defaultRatio
  },
  valueAxis: {
    scale: "linear"
  },
  categoryAxis: {
    tickLabelLayout: "horizontal"
  },
  truncationMode: import_constants.DEFAULT_TRUNCATION_MODE,
  colorPaletteMode: "single-color",
  thresholdIndicators: [],
  toolbar: {
    placement: "top-right",
    hidden: false,
    downloadData: { enabled: false }
  },
  colorRules: [],
  intents: []
};
const DEFAULT_CHART_HEIGHT = 300;
const DEFAULT_LEGEND = {
  position: "auto",
  hidden: false,
  ratio: import_constants.defaultRatio
};
