// @dynatrace/strato-components/charts/timeseries/utils/convert-to-timeseries/utils/create-timeseries-from-record.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var create_timeseries_from_record_exports = {};
__export(create_timeseries_from_record_exports, {
  createTimeseriesFromRecord: () => createTimeseriesFromRecord
});
module.exports = __toCommonJS(create_timeseries_from_record_exports);
var import_check_timeseries_format = require("./check-timeseries-format.js");
var import_extract_start_end_dates = require("./extract-start-end-dates.js");
var import_extractValue = require("./extractValue.js");
var import_get_dimensions = require("./get-dimensions.js");
var import_getFieldsArrayForRecordIndex = require("./getFieldsArrayForRecordIndex.js");
var import_create_name = require("../create-name.js");
function createTimeseriesFromRecord(record, types) {
  const timeseriesArray = [];
  const timeseriesFields = (0, import_check_timeseries_format.checkTimeseriesFormat)(types.mappings);
  if (!timeseriesFields.isTimeseries) {
    return timeseriesArray;
  }
  const fieldsArray = (0, import_getFieldsArrayForRecordIndex.convertToFieldEntries)(types);
  const dimensions = (0, import_get_dimensions.getDimensions)(record, fieldsArray, true);
  const [start] = (0, import_extract_start_end_dates.extractStartEndDates)(record, fieldsArray);
  const valueDimensions = (0, import_get_dimensions.getValueDimensions)(fieldsArray, true);
  for (const valueDimension of valueDimensions ?? []) {
    if (valueDimension.name === "") {
      throw new Error(`Could not calculate key for record ${record}`);
    }
    const timeseries = {
      name: (0, import_create_name.createName)(dimensions, valueDimension.name),
      datapoints: [],
      userData: {
        dimensions: [
          ...dimensions,
          {
            field: {
              name: timeseriesFields.durationField,
              type: "duration"
            },
            value: record[timeseriesFields.durationField]
          },
          {
            field: {
              name: timeseriesFields.timeframeField,
              type: "timeframe"
            },
            value: record[timeseriesFields.timeframeField]
          }
        ],
        valueDimension: valueDimension.name
      }
    };
    timeseriesArray.push(timeseries);
    const values = (0, import_extractValue.extractValue)(
      record,
      valueDimension.name
    );
    const duration = Number(
      BigInt(
        (0, import_extractValue.extractValue)(record, timeseriesFields.durationField)
      ) / BigInt(1e6)
    );
    if (values !== void 0 && values !== null) {
      let newStart = start;
      for (const value of values) {
        const newEnd = new Date(newStart.getTime() + duration);
        if (value !== null) {
          timeseries.datapoints.push({
            start: newStart,
            end: newEnd,
            value: typeof value === "string" ? parseFloat(value) : value
          });
        }
        newStart = newEnd;
      }
    }
  }
  return timeseriesArray;
}
