// @dynatrace/strato-components/charts/timeseries/utils/convert-to-timeseries/utils/unit-mapping.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var unit_mapping_exports = {};
__export(unit_mapping_exports, {
  mapMetricUnitToSdk: () => mapMetricUnitToSdk
});
module.exports = __toCommonJS(unit_mapping_exports);
var import_units = require("@dynatrace-sdk/units");
const mapping = {
  // SI units
  s: import_units.units.time.second,
  m: import_units.units.length.meter,
  kg: import_units.units.mass.kilogram,
  A: import_units.units.electricity.ampere,
  K: import_units.units.temperature.kelvin,
  mol: "mol",
  cd: import_units.units.physics.candela,
  Second: import_units.units.time.second,
  Metre: import_units.units.length.meter,
  Kilogram: import_units.units.mass.kilogram,
  Ampere: import_units.units.electricity.ampere,
  Kelvin: import_units.units.temperature.kelvin,
  Mole: "mol",
  Candela: import_units.units.physics.candela,
  //time
  ns: import_units.units.time.nanosecond,
  us: import_units.units.time.microsecond,
  ms: import_units.units.time.millisecond,
  NanoSecond: import_units.units.time.nanosecond,
  MicroSecond: import_units.units.time.microsecond,
  MilliSecond: import_units.units.time.millisecond,
  min: import_units.units.time.minute,
  h: import_units.units.time.hour,
  d: import_units.units.time.day,
  wk: import_units.units.time.week,
  mo: import_units.units.time.month,
  a: import_units.units.time.year,
  Minute: import_units.units.time.minute,
  Hour: import_units.units.time.hour,
  Day: import_units.units.time.day,
  Week: import_units.units.time.week,
  Month: import_units.units.time.month,
  Year: import_units.units.time.year,
  // time squared
  SquareSecond: import_units.units.time_squared.square_second,
  SquaredSecond: import_units.units.time_squared.square_second,
  s2: import_units.units.time_squared.square_second,
  // length
  "[in_i]": import_units.units.length.inch,
  "[ft_i]": import_units.units.length.foot,
  "[yd_i]": import_units.units.length.yard,
  "[mi_i]": import_units.units.length.mile,
  Inch: import_units.units.length.inch,
  Foot: import_units.units.length.foot,
  Yard: import_units.units.length.yard,
  Mile: import_units.units.length.mile,
  // area
  m2: import_units.units.area.square_meter,
  "[sin_i]": import_units.units.area.square_inch,
  "[sft_i]": import_units.units.area.square_foot,
  "[syd_i]": import_units.units.area.square_yard,
  "[smi_us]": import_units.units.area.square_mile,
  SquareMetre: import_units.units.area.square_meter,
  SquareInch: import_units.units.area.square_inch,
  SquareFoot: import_units.units.area.square_foot,
  SquareYard: import_units.units.area.square_yard,
  SquareMile: import_units.units.area.square_mile,
  // volume
  m3: import_units.units.volume.cubic_meter,
  "[cin_i]": import_units.units.volume.cubic_inch,
  "[cft_i]": import_units.units.volume.cubic_foot,
  l: import_units.units.volume.litre,
  CubicMetre: import_units.units.volume.cubic_meter,
  CubicInch: import_units.units.volume.cubic_inch,
  CubicFoot: import_units.units.volume.cubic_foot,
  Litre: import_units.units.volume.litre,
  // frequency
  Hz: import_units.units.frequency.hertz,
  Hertz: import_units.units.frequency.hertz,
  // mass
  g: import_units.units.mass.gram,
  t: import_units.units.mass.tonne,
  "[lb_av]": import_units.units.mass.pound,
  Gram: import_units.units.mass.gram,
  Tonne: import_units.units.mass.tonne,
  Pound: import_units.units.mass.pound,
  // force
  N: import_units.units.force.newton,
  "[lbf_av]": import_units.units.force.pound,
  Newton: import_units.units.force.newton,
  PoundForce: import_units.units.force.pound,
  // energy
  J: "J",
  Wh: "Wh",
  kWh: "kWh",
  W: import_units.units.electricity.watt,
  Joule: "J",
  WattHour: "Wh",
  KiloWattHour: "kWh",
  Watt: import_units.units.electricity.watt,
  // pressure
  Pa: import_units.units.pressure.pascal,
  kPa: import_units.units.pressure.kilopascal,
  "[lbf_av]/[sin_i]": "psi",
  bar: import_units.units.pressure.bar,
  Pascal: import_units.units.pressure.pascal,
  KiloPascal: import_units.units.pressure.kilopascal,
  PoundForcePerSquareInch: "psi",
  Bar: import_units.units.pressure.bar,
  // electricity
  V: import_units.units.electricity.volt,
  \u03A9: import_units.units.electricity.ohm,
  Volt: import_units.units.electricity.volt,
  Ohm: import_units.units.electricity.ohm,
  // acceleration
  "m/s2": "m/s\xB2",
  "[ft_i]/s2": "ft/s\xB2",
  MetrePerSecondSquared: "m/s\xB2",
  FootPerSecondSquared: "ft/s\xB2",
  "m/s": import_units.units.velocity.meter_per_second,
  MetrePerSecond: import_units.units.velocity.meter_per_second,
  "m/h": "m/h",
  MetrePerHour: "m/h",
  "km/h": import_units.units.velocity.kilometer_per_hour,
  KiloMetrePerHour: import_units.units.velocity.kilometer_per_hour,
  USD: import_units.units.currency.usd,
  EUR: import_units.units.currency.eur,
  JPY: import_units.units.currency.jpy,
  GBP: import_units.units.currency.gbp,
  AUD: import_units.units.currency.aud,
  CAD: import_units.units.currency.cad,
  CHF: import_units.units.currency.chf,
  CNY: import_units.units.currency.cny,
  UnitedStatesDollar: import_units.units.currency.usd,
  Euro: import_units.units.currency.eur,
  BritishPound: import_units.units.currency.gbp,
  AustralianDollar: import_units.units.currency.aud,
  CanadianDollar: import_units.units.currency.cad,
  SwissFranc: import_units.units.currency.chf,
  "chinese yuan": import_units.units.currency.cny,
  // temperature
  Cel: import_units.units.temperature.degree_celsius,
  "[degF]": import_units.units.temperature.degree_fahrenheit,
  "[degR]": import_units.units.temperature.degree_rankine,
  Celsius: import_units.units.temperature.degree_celsius,
  Fahrenheit: import_units.units.temperature.degree_fahrenheit,
  Rankine: import_units.units.temperature.degree_rankine,
  // information
  bit: import_units.units.data.bit,
  kbit: import_units.units.data.kilobit,
  Kibit: import_units.units.data.kibibit,
  Mbit: import_units.units.data.megabit,
  Mibit: import_units.units.data.mebibit,
  Gbit: import_units.units.data.gigabit,
  Gibit: import_units.units.data.gibibit,
  Bit: import_units.units.data.bit,
  KiloBit: import_units.units.data.kilobit,
  KibiBit: import_units.units.data.kibibit,
  MegaBit: import_units.units.data.megabit,
  MebiBit: import_units.units.data.mebibit,
  GigaBit: import_units.units.data.gigabit,
  GibiBit: import_units.units.data.gibibit,
  By: import_units.units.data.byte,
  kBy: import_units.units.data.kilobyte,
  KiBy: import_units.units.data.kibibyte,
  MBy: import_units.units.data.megabyte,
  MiBy: import_units.units.data.mebibyte,
  GBy: import_units.units.data.gigabyte,
  GiBy: import_units.units.data.gibibyte,
  Byte: import_units.units.data.byte,
  KiloByte: import_units.units.data.kilobyte,
  KibiByte: import_units.units.data.kibibyte,
  MegaByte: import_units.units.data.megabyte,
  MebiByte: import_units.units.data.mebibyte,
  GigaByte: import_units.units.data.gigabyte,
  GibiByte: import_units.units.data.gibibyte,
  // information over time
  "By/s": import_units.units.datarate.Bps,
  "By/min": import_units.units.datarate.Bpm,
  "By/h": import_units.units.datarate.Bph,
  "bit/s": import_units.units.datarate.bitps,
  "bit/min": import_units.units.datarate.bitpm,
  "bit/h": import_units.units.datarate.bitph,
  BytePerSecond: import_units.units.datarate.Bps,
  BytePerMinute: import_units.units.datarate.Bpm,
  BytePerHour: import_units.units.datarate.Bph,
  BitPerSecond: import_units.units.datarate.bitps,
  BitPerMinute: import_units.units.datarate.bitpm,
  BitPerHour: import_units.units.datarate.bitph,
  "kbit/s": import_units.units.datarate.kbitps,
  "kBit/min": import_units.units.datarate.kbitpm,
  "kBit/h": import_units.units.datarate.kbitph,
  "kBy/s": import_units.units.datarate.kBps,
  "kBy/min": import_units.units.datarate.kBpm,
  "kBy/h": import_units.units.datarate.kBph,
  "Kibit/s": import_units.units.datarate.kibitps,
  "KiBit/min": import_units.units.datarate.kbitpm,
  "KiBit/h": import_units.units.datarate.kibitph,
  "KiBy/s": import_units.units.datarate.kiBps,
  "KiBy/min": import_units.units.datarate.kiBpm,
  "KiBy/h": import_units.units.datarate.kiBph,
  "Mbit/s": import_units.units.datarate.Mbitps,
  "Mbit/min": import_units.units.datarate.Mbitpm,
  "Mbit/h": import_units.units.datarate.Mbitph,
  "MBy/s": import_units.units.datarate.MBps,
  "MBy/min": import_units.units.datarate.MBpm,
  "MBy/h": import_units.units.datarate.MBph,
  "Mibit/s": import_units.units.datarate.Mibitps,
  "Mibit/min": import_units.units.datarate.Mibitpm,
  "Mibit/h": import_units.units.datarate.Mibitph,
  "MiBy/s": import_units.units.datarate.MiBps,
  "MiBy/min": import_units.units.datarate.MiBpm,
  "MiBy/h": import_units.units.datarate.MiBph,
  "Gbit/s": import_units.units.datarate.Gbitps,
  "Gbit/min": import_units.units.datarate.Gbitpm,
  "Gbit/h": import_units.units.datarate.Gibitph,
  "GBy/s": import_units.units.datarate.GBps,
  "GBy/min": import_units.units.datarate.GBpm,
  "GBy/h": import_units.units.datarate.GBph,
  "Gibit/s": import_units.units.datarate.Gibitps,
  "Gibit/min": import_units.units.datarate.Gibitpm,
  "Gibit/h": import_units.units.datarate.Gibitph,
  "GiBy/s": import_units.units.datarate.GiBps,
  "GiBy/min": import_units.units.datarate.GiBpm,
  "GiBy/h": import_units.units.datarate.GiBph,
  KiloBitPerSecond: import_units.units.datarate.kbitps,
  KiloBitPerMinute: import_units.units.datarate.kbitpm,
  KiloBitPerHour: import_units.units.datarate.kbitph,
  KiloBytePerSecond: import_units.units.datarate.kBps,
  KiloBytePerMinute: import_units.units.datarate.kBpm,
  KiloBytePerHour: import_units.units.datarate.kBph,
  KibiBitPerSecond: import_units.units.datarate.kibitps,
  KibiBitPerMinute: import_units.units.datarate.kbitpm,
  KibiBitPerHour: import_units.units.datarate.kibitph,
  KibiBytePerSecond: import_units.units.datarate.kiBps,
  KibiBytePerMinute: import_units.units.datarate.kiBpm,
  KibiBytePerHour: import_units.units.datarate.kiBph,
  MegaBitPerSecond: import_units.units.datarate.Mbitps,
  MegaBitPerMinute: import_units.units.datarate.Mbitpm,
  MegaBitPerHour: import_units.units.datarate.Mbitph,
  MegaBytePerSecond: import_units.units.datarate.MBps,
  MegaBytePerMinute: import_units.units.datarate.MBpm,
  MegaBytePerHour: import_units.units.datarate.MBph,
  MebiBitPerSecond: import_units.units.datarate.Mibitps,
  MebiBitPerMinute: import_units.units.datarate.Mibitpm,
  MebiBitPerHour: import_units.units.datarate.Mibitph,
  MebiBytePerSecond: import_units.units.datarate.MiBps,
  MebiBytePerMinute: import_units.units.datarate.MiBpm,
  MebiBytePerHour: import_units.units.datarate.MiBph,
  GigaBitPerSecond: import_units.units.datarate.Gbitps,
  GigaBitPerMinute: import_units.units.datarate.Gbitpm,
  GigaBitPerHour: import_units.units.datarate.Gibitph,
  GigaBytePerSecond: import_units.units.datarate.GBps,
  GigaBytePerMinute: import_units.units.datarate.GBpm,
  GigaBytePerHour: import_units.units.datarate.GBph,
  GibiBitPerSecond: import_units.units.datarate.Gibitps,
  GibiBitPerMinute: import_units.units.datarate.Gibitpm,
  GibiBitPerHour: import_units.units.datarate.Gibitph,
  GibiBytePerSecond: import_units.units.datarate.GiBps,
  GibiBytePerMinute: import_units.units.datarate.GiBpm,
  GibiBytePerHour: import_units.units.datarate.GiBph,
  // count
  count: "count",
  "s/min": "s/min",
  Count: "count",
  // count over time
  "count/s": "count/s",
  "count/min": "count/min",
  "count/h": "count/h",
  "[ppth]": import_units.units.percentage.per_mille,
  "ms/min": "ms/min",
  "ns/min": "ms/min",
  MilliSecondPerMinute: "ms/min",
  NanoSecondPerMinute: "ns/min",
  PerSecond: "count/s",
  PerMinute: "count/min",
  PerHour: "count/h",
  core: "core",
  Cores: "core",
  MilliCores: "mcore",
  px: "px",
  Pixel: "px",
  msu: "msu",
  MSU: "msu",
  ratio: "ratio",
  Ratio: "ratio",
  Percent: import_units.units.percentage.percent,
  "%": import_units.units.percentage.percent,
  Promille: import_units.units.percentage.per_mille,
  state: "state",
  State: "state",
  Decibel: "dB",
  dBm: "dBm",
  DecibelMilliWatt: "dBm",
  Unspecified: "unspecified"
};
function mapMetricUnitToSdk(unitId) {
  if (unitId === "Unspecified") {
    return void 0;
  }
  return mapping[unitId] ?? unitId;
}
