// @dynatrace/strato-components/charts/tree-map/providers/TreeMapChartProviders.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TreeMapChartProviders_exports = {};
__export(TreeMapChartProviders_exports, {
  TreeMapChartProviders: () => TreeMapChartProviders
});
module.exports = __toCommonJS(TreeMapChartProviders_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_StoreProvider = require("./StoreProvider.js");
var import_ChartSeriesActionContext = require("../../core/components/series-action/ChartSeriesActionContext.js");
var import_ColorRuleContext = require("../../core/context/ColorRuleContext.js");
var import_ChartLegendContext = require("../context/ChartLegendContext.js");
var import_ColorPaletteContext = require("../context/ColorPaletteContext.js");
var import_FormatterContext = require("../context/FormatterContext.js");
var import_LabelsDisplayContext = require("../context/LabelsDisplayContext.js");
var import_MinMaxConfigContext = require("../context/MinMaxConfigContext.js");
var import_NameAccessorContext = require("../context/NameAccessorContext.js");
var import_ToolbarContext = require("../context/ToolbarContext.js");
var import_TreeMapRawDataContext = require("../context/TreeMapRawDataContext.js");
var import_TreeMapCustomTooltipContext = require("../contexts/TreeMapCustomTooltipContext.js");
const TreeMapChartProviders = (props) => {
  const {
    data,
    seriesActions,
    formatter,
    truncationMode,
    legend,
    toolbar,
    labelsDisplay,
    nameAccessor,
    colorPalette,
    min: configMin,
    max: configMax,
    children,
    colorRules = [],
    unit,
    tooltip
  } = props;
  const min = configMin === "data-min" ? void 0 : configMin;
  const max = configMax === "data-max" ? void 0 : configMax;
  const memoizedRawData = (0, import_react.useMemo)(
    () => ({ ...data, unit: unit ?? data.unit }),
    [data, unit]
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ColorRuleContext.ColorRuleContext.Provider, { value: colorRules, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TreeMapRawDataContext.TreeMapRawDataContext.Provider, { value: memoizedRawData, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ChartSeriesActionContext.ChartSeriesActionContext.Provider,
    {
      value: {
        // eslint-disable-next-line @typescript-eslint/prefer-ts-expect-error, @typescript-eslint/ban-ts-comment
        // @ts-ignore Strict mode transition - needs to be investigated
        chartSeriesActions: seriesActions
      },
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_FormatterContext.FormatterContext.Provider,
        {
          value: {
            formatter,
            truncationMode
          },
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ColorPaletteContext.ColorPaletteContext.Provider, { value: colorPalette, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ChartLegendContext.ChartLegendContext.Provider, { value: legend, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_NameAccessorContext.NameAccessorContext.Provider, { value: nameAccessor, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ToolbarContext.ToolbarContext.Provider, { value: toolbar, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_LabelsDisplayContext.LabelsDisplayContext.Provider, { value: labelsDisplay, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_MinMaxConfigContext.MinMaxConfigContext.Provider, { value: { min, max }, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_TreeMapCustomTooltipContext.TreeMapCustomTooltipContext.Provider,
            {
              value: tooltip ?? import_TreeMapCustomTooltipContext.DEFAULT_CUSTOM_TOOLTIP_CONFIG,
              children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_StoreProvider.StoreProvider, { children })
            }
          ) }) }) }) }) }) })
        }
      )
    }
  ) }) });
};
TreeMapChartProviders["displayName"] = "TreeMapChartProviders";
