// @dynatrace/strato-components/esm/charts/annotations-chart/components/axis/AnnotationsAxis.js
import { jsx } from "react/jsx-runtime";
import { useAnnotationsAxisConfig } from "./hooks/useAnnotationsAxisConfig.js";
import {
  isNumericalAxis,
  isNumericalScale
} from "./types/annotations-x-axis-props.js";
import { TimeBottomAxis } from "../../../core/components/axes/components/TimeBottomAxis.js";
import { ValueBottomAxis } from "../../../core/components/axes/components/ValueBottomAxis.js";
import { getFormattedTicks } from "../../../core/utils/formatter-by-ticks.js";
import { getFormatter } from "../../../core/utils/get-formatter.js";
import { getHorizontalNumTicks } from "../../../core/utils/ticks.js";
import { DEFAULT_AXIS_HIDDEN } from "../../constants.js";
const AnnotationsAxis = ({
  scale,
  width,
  height,
  labelsOffset
}) => {
  const config = useAnnotationsAxisConfig();
  const horizontalNumTicks = getHorizontalNumTicks(width);
  const hidden = config?.hidden ?? DEFAULT_AXIS_HIDDEN;
  let formatterResult;
  if (config && isNumericalAxis(config)) {
    const formatter = getFormatter(config.formatter);
    formatterResult = getFormattedTicks(
      horizontalNumTicks,
      scale,
      config.unit,
      formatter
    );
  } else {
    formatterResult = getFormattedTicks(
      horizontalNumTicks,
      scale,
      void 0,
      void 0
    );
  }
  const commonProps = {
    containerWidth: width - labelsOffset,
    containerHeight: height,
    ticks: formatterResult.ticks,
    formattedTicks: formatterResult.formattedTicks,
    label: config?.label
  };
  return !hidden ? /* @__PURE__ */ jsx(
    "div",
    {
      style: {
        marginLeft: `${labelsOffset}px`
      },
      children: isNumericalScale(scale) ? /* @__PURE__ */ jsx(ValueBottomAxis, { scale, ...commonProps }) : /* @__PURE__ */ jsx(TimeBottomAxis, { scale, ...commonProps })
    }
  ) : null;
};
export {
  AnnotationsAxis
};
