// @dynatrace/strato-components/esm/charts/timeseries/utils/convert-to-timeseries/convert-to-timeseries-band.js
import { createTimeseriesFromRecord } from "./utils/create-timeseries-from-record.js";
const upperSuffix = ["max", "upper"];
const lowerSuffix = ["min", "lower"];
function valueDimensionEndsWith(timeseries, suffixList) {
  return suffixList.some(
    (suffix) => timeseries.userData.valueDimension.endsWith(suffix)
  );
}
function getDimensionsWithoutLast(name) {
  if (typeof name === "string") {
    return "band";
  }
  const timeseriesName = name.slice(0, -1);
  return timeseriesName.length > 0 ? timeseriesName : "band";
}
function convertToTimeseriesBand(record, types) {
  const timeseries = createTimeseriesFromRecord(record, types);
  const upper = timeseries.find(
    (timeseries2) => valueDimensionEndsWith(timeseries2, upperSuffix)
  );
  const lower = timeseries.find(
    (timeseries2) => valueDimensionEndsWith(timeseries2, lowerSuffix)
  );
  const rest = timeseries.filter(
    (timeseries2) => timeseries2 !== lower && timeseries2 !== upper
  );
  if (upper && lower && upper.datapoints.length === lower.datapoints.length) {
    const timeseriesBand = {
      name: getDimensionsWithoutLast(upper.name),
      datapoints: upper.datapoints.map(
        ({ value, ...datapoint }, index) => ({
          ...datapoint,
          y0: lower?.datapoints[index].value,
          y1: value
        })
      ),
      userData: {
        ...upper.userData,
        lowerValueDimension: lower.userData.valueDimension,
        upperValueDimension: upper.userData.valueDimension
      }
    };
    return {
      timeseriesBand,
      timeseries: rest,
      timeseriesOfBand: [upper, lower]
    };
  }
}
export {
  convertToTimeseriesBand
};
