// @dynatrace/strato-components/esm/charts/timeseries/utils/convert-to-timeseries/convert-to-timeseries.js
import { units } from "@dynatrace-sdk/units";
import { createName } from "./create-name.js";
import { createTimeseriesFromRecord } from "./utils/create-timeseries-from-record.js";
import { extractStartEndDates } from "./utils/extract-start-end-dates.js";
import { extractValue } from "./utils/extractValue.js";
import {
  getDimensions,
  getValueDimensions
} from "./utils/get-dimensions.js";
import { _getMetadataMetricsUnitForFieldName } from "./utils/get-metadata-metrics-unit-for-field-name.js";
import {
  getFieldsArrayForRecordIndex,
  getTypesForRecordIndex
} from "./utils/getFieldsArrayForRecordIndex.js";
import { mapMetricUnitToSdk } from "./utils/unit-mapping.js";
function createKeyFromTimeseries(name) {
  if (typeof name === "string") {
    return name;
  }
  const dimensionKeys = [...name];
  const last = dimensionKeys.pop();
  const baseKey = dimensionKeys.join("-");
  return baseKey ? `${baseKey}_${last}` : last ?? "";
}
function convertQueryResultToTimeseries(queryResult) {
  if (queryResult.records === null) {
    return [];
  }
  return convertToTimeseries(
    queryResult.records,
    queryResult.types,
    void 0,
    queryResult.metadata
  );
}
function convertToTimeseries(records, fieldTypes, hiddenFieldIds, metadata) {
  const timeseriesMap = /* @__PURE__ */ new Map();
  const valueDimensionSet = /* @__PURE__ */ new Set();
  for (const [recordIndex, record] of records.entries()) {
    if (record === null) {
      continue;
    }
    const types = getTypesForRecordIndex(fieldTypes, recordIndex);
    if (!types || !record) {
      break;
    }
    const timeseriesArray = createTimeseriesFromRecord(record, types);
    if (timeseriesArray.length > 0) {
      timeseriesArray.forEach((timeseries2) => {
        const valueDimension = timeseries2.userData.valueDimension;
        valueDimensionSet.add(valueDimension);
        if (!hiddenFieldIds?.includes(valueDimension)) {
          const unit = _getMetadataMetricsUnitForFieldName(
            metadata,
            valueDimension
          );
          const newTimeseries = {
            ...timeseries2,
            unit: unit !== void 0 ? mapMetricUnitToSdk(unit) : void 0
          };
          const mapKey = createKeyFromTimeseries(timeseries2.name);
          const uniqueMapKeyAcrossRecords = JSON.stringify([
            mapKey,
            recordIndex
          ]);
          timeseriesMap.set(uniqueMapKeyAcrossRecords, newTimeseries);
        }
      });
      continue;
    }
    const fieldsArray = getFieldsArrayForRecordIndex(fieldTypes, recordIndex);
    let dimensions = getDimensions(record, fieldsArray, false);
    if (hiddenFieldIds && hiddenFieldIds.length) {
      dimensions = dimensions.filter((dimension) => {
        return !hiddenFieldIds.includes(dimension.field.name);
      });
    }
    const [start, end] = extractStartEndDates(record, fieldsArray);
    const valueDimensions = getValueDimensions(fieldsArray, false);
    for (const valueDimension of valueDimensions ?? []) {
      const name = createName(dimensions, valueDimension.name);
      valueDimensionSet.add(valueDimension.name);
      if (name.length === 0) {
        throw new Error(
          `Could not calculate key for record ${JSON.stringify(record)}`
        );
      }
      if (hiddenFieldIds?.includes(valueDimension.name)) {
        continue;
      }
      const timeseries2 = ensureMapEntry(
        timeseriesMap,
        name,
        dimensions,
        valueDimension.name
      );
      const rawValue = extractValue(
        record,
        valueDimension.name
      );
      const value = typeof rawValue === "string" ? parseFloat(rawValue) : rawValue;
      if (valueDimension.type === "duration") {
        timeseries2.unit = units.time.nanosecond;
      }
      if (value !== void 0) {
        timeseries2.datapoints.push({ start, end, value });
      }
    }
  }
  const timeseries = Array.from(timeseriesMap.values()).filter(
    (timeseries2) => timeseries2.datapoints.length > 0
  );
  timeseries.forEach((t) => {
    t.datapoints.sort((a, b) => a.start.getTime() - b.start.getTime());
    if (valueDimensionSet.size === 1 && Array.isArray(t.name) && t.name.length > 1) {
      t.name.pop();
    }
  });
  return timeseries;
}
function ensureMapEntry(timeseriesMap, name, dimensions, valueDimension) {
  const key = createKeyFromTimeseries(name);
  if (key !== void 0 && key !== null && !timeseriesMap.has(key)) {
    const timeseries = {
      datapoints: [],
      name,
      userData: {
        dimensions,
        valueDimension
      }
    };
    timeseriesMap.set(key, timeseries);
    return timeseries;
  }
  return timeseriesMap.get(key);
}
export {
  convertQueryResultToTimeseries,
  convertToTimeseries
};
