// @dynatrace/strato-components/esm/charts/timeseries/components/axes/TimeseriesValueLeftAxisArea.js
import { jsx } from "react/jsx-runtime";
import { ValueLeftAxis } from "../../../core/components/axes/components/ValueLeftAxis.js";
import { ChartGrid } from "../../../core/components/chart-grid/ChartGridRoot.js";
import { useTimeseriesFilteredYAxis } from "../../hooks/useTimeseriesFilteredYAxis.js";
import { useTimeseriesTickAxis } from "../../hooks/useTimeseriesTickAxis.js";
import { useYScales } from "../../hooks/useYScales.js";
import { AREAS } from "../../types/timeseries-grid-area-rules.js";
const TimeseriesValueLeftAxisArea = () => {
  const { left: axisLeftConfig } = useTimeseriesFilteredYAxis();
  const { left: scaleLeft } = useYScales();
  const tickAxis = useTimeseriesTickAxis();
  const leftTickAxis = tickAxis?.left || { ticks: [], formattedTicks: [] };
  if (!leftTickAxis.ticks.length || !scaleLeft) {
    return null;
  }
  return /* @__PURE__ */ jsx(ChartGrid.GridArea, { gridArea: AREAS.leftAxis, children: ({ availableWidth, availableHeight }) => {
    return /* @__PURE__ */ jsx(
      ValueLeftAxis,
      {
        ticks: leftTickAxis.ticks,
        formattedTicks: leftTickAxis.formattedTicks,
        containerHeight: availableHeight,
        containerWidth: availableWidth,
        scale: scaleLeft.scale,
        label: axisLeftConfig?.label
      }
    );
  } });
};
TimeseriesValueLeftAxisArea["displayName"] = "TimeseriesValueLeftAxisArea";
export {
  TimeseriesValueLeftAxisArea
};
