// @dynatrace/strato-components/charts/xy-chart/components/displays/AreaDisplay.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var AreaDisplay_exports = {};
__export(AreaDisplay_exports, {
  AreaDisplay: () => AreaDisplay
});
module.exports = __toCommonJS(AreaDisplay_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_area_accessors = require("./area-accessors.js");
var import_band_accessors = require("./band-accessors.js");
var import_AreaCanvas = require("../../../core/components/canvas/components/AreaCanvas.js");
var import_display_points = require("../../../core/components/canvas/utils/display-points.js");
var import_useGridPlotSize = require("../../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_constants = require("../../../core/components/data-point/constants.js");
var import_constants2 = require("../../../core/constants.js");
var import_is_isolated_point = require("../../../core/utils/is-isolated-point.js");
var import_defaults = require("../../constants/defaults.js");
var import_useXYChartData = require("../../hooks/useXYChartData.js");
var import_get_scale = require("../../utils/get-scale.js");
var import_validation_datapoint = require("../../utils/validation-datapoint.js");
var import_value_representation = require("../../utils/value-representation.js");
function AreaDisplay({
  axisIdToScales,
  series,
  metadata
}) {
  const { height, width } = (0, import_useGridPlotSize.useGridPlotSize)();
  const { datapointMetadataMap, axisIdToScaleDomain } = (0, import_useXYChartData.useXYChartData)();
  const xScale = (0, import_get_scale.getCurrentScale)(axisIdToScales, series.xAxisId);
  const yScale = (0, import_get_scale.getCurrentScale)(axisIdToScales, series.yAxisId);
  const {
    curve,
    color,
    orientation,
    variant,
    pointsDisplay,
    xAxis,
    yAxis,
    isStacked
  } = metadata;
  const xRelAccessor = (0, import_value_representation.relAccessor)(xAxis, axisIdToScaleDomain);
  const yRelAccessor = (0, import_value_representation.relAccessor)(yAxis, axisIdToScaleDomain);
  const showBandedArea = variant === "band";
  const definedAccessor = (0, import_validation_datapoint.isGapDatapointForDisplay)(datapointMetadataMap);
  const showDataPoints = variant === "band" ? pointsDisplay === "always" : (0, import_display_points.shouldDisplayPoints)(
    pointsDisplay ?? import_defaults.DEFAULT_POINTS_DISPLAY,
    width,
    series.data.length
  );
  const isolatedAccessor = (datapoints, index) => showDataPoints || (0, import_is_isolated_point.isIsolatedPoint)(
    datapoints,
    index,
    (point) => (0, import_validation_datapoint.isGapDatapointForDisplay)(datapointMetadataMap)(point)
  );
  const isHorizontal = orientation === "horizontal";
  let primaryAxisAccessor;
  let baselineAccessor;
  let endpointAccessor;
  if (variant === "band") {
    primaryAxisAccessor = isHorizontal ? (0, import_band_accessors.bandY0Accessor)(yScale, yRelAccessor) : (0, import_band_accessors.bandX0Accessor)(xScale, xRelAccessor);
    baselineAccessor = isHorizontal ? (0, import_band_accessors.bandX0Accessor)(xScale, xRelAccessor) : (0, import_band_accessors.bandY0Accessor)(yScale, yRelAccessor);
    endpointAccessor = isHorizontal ? (0, import_band_accessors.bandX1Accessor)(xScale, xRelAccessor) : (0, import_band_accessors.bandY1Accessor)(yScale, yRelAccessor);
  } else {
    primaryAxisAccessor = isHorizontal ? (0, import_area_accessors.yAccessor)(yScale, yRelAccessor) : (0, import_area_accessors.xAccessor)(xScale, xRelAccessor);
    baselineAccessor = isHorizontal ? (0, import_area_accessors.x0Accessor)(xScale, datapointMetadataMap, xRelAccessor, isStacked) : (0, import_area_accessors.y0Accessor)(yScale, datapointMetadataMap, yRelAccessor, isStacked);
    endpointAccessor = isHorizontal ? (0, import_area_accessors.x1Accessor)(xScale, datapointMetadataMap, xRelAccessor, isStacked) : (0, import_area_accessors.y1Accessor)(yScale, datapointMetadataMap, yRelAccessor, isStacked);
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_AreaCanvas.AreaCanvas,
    {
      data: series.data,
      canvasWidth: width,
      canvasHeight: height,
      opacity: import_constants2.DEFAULT_FILL_OPACITY,
      color: () => color.Default,
      x: primaryAxisAccessor,
      y0: baselineAccessor,
      y1: endpointAccessor,
      defined: definedAccessor,
      displayScatterPlot: isolatedAccessor,
      enableDotOverflow: true,
      overlayPadding: import_constants.POINT_OVERLAY_PADDING,
      orientation,
      curve,
      showBandedArea
    },
    series.id
  );
}
AreaDisplay["displayName"] = "AreaDisplay";
