// @dynatrace/strato-components/charts/xy-chart/components/displays/area-accessors.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var area_accessors_exports = {};
__export(area_accessors_exports, {
  applyNumericalBarScaleTransformations: () => applyNumericalBarScaleTransformations,
  x0Accessor: () => x0Accessor,
  x1Accessor: () => x1Accessor,
  xAccessor: () => xAccessor,
  y0Accessor: () => y0Accessor,
  y1Accessor: () => y1Accessor,
  yAccessor: () => yAccessor
});
module.exports = __toCommonJS(area_accessors_exports);
var import_lodash_es = require("lodash");
var import_get_band_centered_point = require("../../utils/get-band-centered-point.js");
var import_get_center_date = require("../../utils/get-center-date.js");
var import_type_guards = require("../../utils/type-guards.js");
function applyNumericalBarScaleTransformations(args, accessorCoord, baseline, relAccessor, accumulator, binDomain, isSeriesStacked) {
  const isNotBaseline = baseline !== void 0 && accessorCoord !== baseline;
  let { v0, v1 } = args;
  if (isNotBaseline && accumulator !== void 0) {
    v0 = !(0, import_lodash_es.isNil)(v0) ? v0 + accumulator : accumulator;
    v1 = !(0, import_lodash_es.isNil)(v1) ? v1 + accumulator : accumulator;
  }
  if (relAccessor && isNotBaseline && !(0, import_lodash_es.isNil)(binDomain) && isSeriesStacked) {
    return {
      v0: !(0, import_lodash_es.isNil)(v0) ? v0 / binDomain * 100 : v0,
      v1: !(0, import_lodash_es.isNil)(v1) ? v1 / binDomain * 100 : v1,
      scale: args.scale
    };
  }
  if (relAccessor) {
    return {
      v0: !(0, import_lodash_es.isNil)(v0) ? relAccessor(v0) : v0,
      v1: !(0, import_lodash_es.isNil)(v1) ? relAccessor(v1) : v1,
      scale: args.scale
    };
  }
  return { v0, v1, scale: args.scale };
}
const getPrimaryAxisAccessor = (accessorKeys) => {
  return (scale, relAccessor) => {
    return (datapoint) => {
      const v0 = datapoint[accessorKeys[0]];
      const v1 = datapoint[accessorKeys[1]];
      if (v0 === void 0) {
        throw Error("unknown Accessor value for Area series");
      }
      const args = { scale, v0, v1 };
      if ((0, import_type_guards.isNumericalAreaScale)(args)) {
        return args.scale(relAccessor ? relAccessor(args.v0) : args.v0);
      }
      if ((0, import_type_guards.isTimeAreaScale)(args)) {
        const value = (0, import_get_center_date.getCenterDate)(args.v0, args.v1);
        return args.scale(value);
      }
      if ((0, import_type_guards.isCategoricalAreaScale)(args)) {
        return (0, import_get_band_centered_point.getBandCenteredPoint)(args.scale, args.v0);
      }
      throw Error("unknown Accessor value for Area series");
    };
  };
};
const getGrowingAccessors = (accessorKeys, baselineCoord) => {
  const baselineAccessor = (scale, datapointsMetadata, relAccessor, isSeriesStacked) => {
    return (datapoint) => {
      const v0 = datapoint[accessorKeys[0]];
      const v1 = datapoint[accessorKeys[1]];
      const { accumulator, baseline, binDomain } = datapointsMetadata.get(datapoint) ?? {};
      let args = { scale, v0, v1 };
      if ((0, import_type_guards.isNumericalAreaScale)(args)) {
        args = applyNumericalBarScaleTransformations(
          args,
          baselineCoord === "x" ? "y" : "x",
          baseline,
          relAccessor,
          accumulator,
          binDomain,
          isSeriesStacked
        );
      }
      if ((0, import_type_guards.isNumericalAreaScale)(args)) {
        const value = args.v0 ?? 0;
        return args.scale(value);
      }
      if ((0, import_type_guards.isTimeAreaScale)(args)) {
        return args.scale(args.scale.invert(0));
      }
      if ((0, import_type_guards.isCategoricalAreaScale)(args)) {
        return args.scale.range()[1];
      }
      throw Error("unknown baseline accessor for Area series");
    };
  };
  const endpointAccessor = (scale, datapointsMetadata, relAccessor, isSeriesStacked) => {
    return (datapoint) => {
      const v0 = datapoint[accessorKeys[0]];
      const v1 = datapoint[accessorKeys[1]];
      const { accumulator, baseline, binDomain } = datapointsMetadata.get(datapoint) ?? {};
      if (v1 === void 0) {
        throw Error(
          `${accessorKeys[1]} cannot be undefined for the growing accessor in the AreaSeries`
        );
      }
      let args = { scale, v0, v1 };
      if ((0, import_type_guards.isNumericalAreaScale)(args)) {
        args = applyNumericalBarScaleTransformations(
          args,
          baselineCoord === "x" ? "y" : "x",
          baseline,
          relAccessor,
          accumulator,
          binDomain,
          isSeriesStacked
        );
      }
      if ((0, import_type_guards.isTimeAreaScale)(args) || (0, import_type_guards.isNumericalAreaScale)(args)) {
        return args.scale(args.v1);
      }
      if ((0, import_type_guards.isCategoricalAreaScale)(args)) {
        const catWidth = Math.floor(args.scale.bandwidth() / 2);
        return (args.scale(args.v1) ?? 0) + catWidth;
      }
      throw Error("unknown value accessor for Area series");
    };
  };
  return { baselineAccessor, endpointAccessor };
};
const xAccessor = (scale, relAccessor) => getPrimaryAxisAccessor(["x0", "x1"])(scale, relAccessor);
const { baselineAccessor: y0Accessor, endpointAccessor: y1Accessor } = getGrowingAccessors(["y0", "y1"], "x");
const yAccessor = (scale, relAccessor) => getPrimaryAxisAccessor(["y0", "y1"])(scale, relAccessor);
const { baselineAccessor: x0Accessor, endpointAccessor: x1Accessor } = getGrowingAccessors(["x0", "x1"], "y");
