// @dynatrace/strato-components/charts/xy-chart/components/displays/band-accessors.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var band_accessors_exports = {};
__export(band_accessors_exports, {
  bandX0Accessor: () => bandX0Accessor,
  bandX1Accessor: () => bandX1Accessor,
  bandY0Accessor: () => bandY0Accessor,
  bandY1Accessor: () => bandY1Accessor,
  xMiddleBandAccessor: () => xMiddleBandAccessor,
  yMiddleBandAccessor: () => yMiddleBandAccessor
});
module.exports = __toCommonJS(band_accessors_exports);
var import_get_band_centered_point = require("../../utils/get-band-centered-point.js");
var import_get_center_date = require("../../utils/get-center-date.js");
var import_type_guards = require("../../utils/type-guards.js");
const getBandValueAccessor = (accessorKey) => {
  return (scale, relAccessor) => {
    return (datapoint) => {
      const value = datapoint[accessorKey];
      if (value === void 0) {
        throw Error(
          `${accessorKey} cannot be undefined for band chart accessor`
        );
      }
      const args = { scale, v0: value, v1: void 0 };
      if ((0, import_type_guards.isNumericalAreaScale)(args)) {
        return args.scale(relAccessor ? relAccessor(args.v0) : args.v0);
      }
      if ((0, import_type_guards.isTimeAreaScale)(args)) {
        const isXAccessor = accessorKey === "x0" || accessorKey === "x1";
        const hasAllBandCoordinates = datapoint.x0 !== void 0 && datapoint.x1 !== void 0 && datapoint.y0 !== void 0 && datapoint.y1 !== void 0;
        if (hasAllBandCoordinates && isXAccessor) {
          const centeredTimestamp = (0, import_get_center_date.getCenterDate)(
            datapoint.x0,
            datapoint.x1
          );
          return args.scale(new Date(centeredTimestamp));
        }
        return args.scale(args.v0);
      }
      if ((0, import_type_guards.isCategoricalAreaScale)(args)) {
        return (0, import_get_band_centered_point.getBandCenteredPoint)(args.scale, args.v0);
      }
      throw Error(`unknown scale type for band accessor ${accessorKey}`);
    };
  };
};
const bandX0Accessor = (scale, relAccessor) => getBandValueAccessor("x0")(scale, relAccessor);
const bandX1Accessor = (scale, relAccessor) => getBandValueAccessor("x1")(scale, relAccessor);
const bandY0Accessor = (scale, relAccessor) => getBandValueAccessor("y0")(scale, relAccessor);
const bandY1Accessor = (scale, relAccessor) => getBandValueAccessor("y1")(scale, relAccessor);
function xMiddleBandAccessor(scale, relAccessor) {
  return (datapoint) => {
    const x0 = bandX0Accessor(scale, relAccessor)(datapoint);
    const x1 = bandX1Accessor(scale, relAccessor)(datapoint);
    return (x0 + x1) / 2;
  };
}
function yMiddleBandAccessor(scale, relAccessor) {
  return (datapoint) => {
    const y0 = bandY0Accessor(scale, relAccessor)(datapoint);
    const y1 = bandY1Accessor(scale, relAccessor)(datapoint);
    return (y0 + y1) / 2;
  };
}
