// @dynatrace/strato-components/charts/xy-chart/components/displays/DotDisplay.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DotDisplay_exports = {};
__export(DotDisplay_exports, {
  DotDisplay: () => DotDisplay
});
module.exports = __toCommonJS(DotDisplay_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_dot_accessors = require("./dot-accessors.js");
var import_DotCanvas = require("../../../core/components/canvas/components/DotCanvas.js");
var import_useGridPlotSize = require("../../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_constants = require("../../../core/constants.js");
var import_useXYChartData = require("../../hooks/useXYChartData.js");
var import_get_scale = require("../../utils/get-scale.js");
var import_value_representation = require("../../utils/value-representation.js");
function DotDisplay({
  axisIdToScales,
  series,
  metadata
}) {
  const { height, width } = (0, import_useGridPlotSize.useGridPlotSize)();
  const { axisIdToScaleDomain } = (0, import_useXYChartData.useXYChartData)();
  const xScale = (0, import_get_scale.getCurrentScale)(axisIdToScales, series.xAxisId);
  const yScale = (0, import_get_scale.getCurrentScale)(axisIdToScales, series.yAxisId);
  const { xAxis, yAxis } = metadata;
  const xRelAcc = (0, import_value_representation.relAccessor)(xAxis, axisIdToScaleDomain);
  const yRelAcc = (0, import_value_representation.relAccessor)(yAxis, axisIdToScaleDomain);
  const { color, shape } = metadata;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_DotCanvas.DotCanvas,
    {
      data: series.data,
      canvasHeight: height,
      canvasWidth: width,
      color: () => color.Default,
      opacity: import_constants.DEFAULT_FILL_OPACITY,
      x: (0, import_dot_accessors.xAccessor)(xScale, xRelAcc),
      y: (0, import_dot_accessors.yAccessor)(yScale, yRelAcc),
      shape,
      enableDotOverflow: true
    },
    series.id
  );
}
DotDisplay["displayName"] = "DotDisplay";
