// @dynatrace/strato-components/charts/xy-chart/components/displays/LineDisplay.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var LineDisplay_exports = {};
__export(LineDisplay_exports, {
  LineDisplay: () => LineDisplay
});
module.exports = __toCommonJS(LineDisplay_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_line_accessors = require("./line-accessors.js");
var import_LineCanvas = require("../../../core/components/canvas/components/LineCanvas.js");
var import_display_points = require("../../../core/components/canvas/utils/display-points.js");
var import_useGridPlotSize = require("../../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_constants = require("../../../core/components/data-point/constants.js");
var import_constants2 = require("../../../core/constants.js");
var import_is_isolated_point = require("../../../core/utils/is-isolated-point.js");
var import_defaults = require("../../constants/defaults.js");
var import_useXYChartData = require("../../hooks/useXYChartData.js");
var import_get_scale = require("../../utils/get-scale.js");
var import_validation_datapoint = require("../../utils/validation-datapoint.js");
var import_value_representation = require("../../utils/value-representation.js");
function LineDisplay({
  axisIdToScales,
  series,
  metadata
}) {
  const { height, width } = (0, import_useGridPlotSize.useGridPlotSize)();
  const { datapointMetadataMap, axisIdToScaleDomain } = (0, import_useXYChartData.useXYChartData)();
  const xScale = (0, import_get_scale.getCurrentScale)(axisIdToScales, series.xAxisId);
  const yScale = (0, import_get_scale.getCurrentScale)(axisIdToScales, series.yAxisId);
  const { xAxis, yAxis } = metadata;
  const xRelAccessor = (0, import_value_representation.relAccessor)(xAxis, axisIdToScaleDomain);
  const yRelAccessor = (0, import_value_representation.relAccessor)(yAxis, axisIdToScaleDomain);
  const { curve, pointsDisplay, color } = metadata;
  const showDataPoints = (0, import_display_points.shouldDisplayPoints)(
    pointsDisplay ?? import_defaults.DEFAULT_POINTS_DISPLAY,
    width,
    series.data.length
  );
  const isolatedAccessor = (datapoints, index) => showDataPoints || (0, import_is_isolated_point.isIsolatedPoint)(
    datapoints,
    index,
    (point) => (0, import_validation_datapoint.isGapDatapointForDisplay)(datapointMetadataMap)(point)
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_LineCanvas.LineCanvas,
    {
      data: series.data,
      canvasHeight: height,
      canvasWidth: width,
      opacity: import_constants2.DEFAULT_FILL_OPACITY,
      color: () => color.Default,
      x: (0, import_line_accessors.xAccessor)(xScale, xRelAccessor),
      y: (0, import_line_accessors.yAccessor)(yScale, yRelAccessor),
      enableDotOverflow: true,
      displayScatterPlot: isolatedAccessor,
      overlayPadding: import_constants.POINT_OVERLAY_PADDING,
      curve,
      defined: (0, import_validation_datapoint.isGapDatapointForDisplay)(datapointMetadataMap)
    },
    series.id
  );
}
LineDisplay["displayName"] = "LineDisplay";
