// @dynatrace/strato-components/charts/xy-chart/components/displays/rect-accessors.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var rect_accessors_exports = {};
__export(rect_accessors_exports, {
  colorAccessor: () => colorAccessor,
  heightAccessor: () => heightAccessor,
  widthAccessor: () => widthAccessor,
  xAccessor: () => xAccessor,
  yAccessor: () => yAccessor
});
module.exports = __toCommonJS(rect_accessors_exports);
var import_safe_get = require("../../../core/utils/safe-get.js");
var import_get_color_from_scale_and_data = require("../../utils/coloring/get-color-from-scale-and-data.js");
var import_type_guards = require("../../utils/type-guards.js");
function colorAccessor(colorScale, metaDataMap) {
  return (datapoint) => (
    // At this point we know value can not be null, undefined nor NaN, hence it will not be painted in the canvas
    (0, import_get_color_from_scale_and_data.getColorFromScaleAndData)(
      colorScale,
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      datapoint.value,
      (0, import_safe_get.safeGet)(metaDataMap, datapoint)?.colorOverride
    )
  );
}
function genericAccessor(scale, v0, v1, relAccessor) {
  const args = { scale, v0, v1 };
  if ((0, import_type_guards.isCategoricalRectScale)(args)) {
    return args.scale(args.v0) ?? 0;
  } else if ((0, import_type_guards.isNumericalRectScale)(args) || (0, import_type_guards.isTimeRectScale)(args)) {
    const { scale: scale2 } = args;
    let { v0: v02, v1: v12 } = args;
    if (relAccessor) {
      if ((0, import_type_guards.isNumericalRectScale)(args)) {
        v02 = relAccessor(v02);
        v12 = relAccessor(v12);
      }
    }
    return Math.min(scale2(v02), scale2(v12));
  }
  throw Error("unknown genericAccessor value for Rect");
}
function spanAccessor(scale, v0, v1, relAccessor) {
  const args = { scale, v0, v1 };
  if ((0, import_type_guards.isNumericalRectScale)(args) || (0, import_type_guards.isTimeRectScale)(args)) {
    const { scale: scale2 } = args;
    let { v0: v02, v1: v12 } = args;
    if (relAccessor) {
      if ((0, import_type_guards.isNumericalRectScale)(args)) {
        v02 = relAccessor(v02);
        v12 = relAccessor(v12);
      }
    }
    return Math.abs(scale2(v12) - scale2(v02));
  } else if ((0, import_type_guards.isCategoricalRectScale)(args)) {
    return args.scale.bandwidth();
  }
  throw Error("unknown spanAccessor value for Rect");
}
function xAccessor(scale, relAccessor) {
  return (datapoint) => genericAccessor(scale, datapoint.x0, datapoint.x1, relAccessor);
}
function yAccessor(scale, relAccessor) {
  return (datapoint) => genericAccessor(scale, datapoint.y0, datapoint.y1, relAccessor);
}
function widthAccessor(scale, relAccessor) {
  return (datapoint) => spanAccessor(scale, datapoint.x0, datapoint.x1, relAccessor);
}
function heightAccessor(scale, relAccessor) {
  return (datapoint) => spanAccessor(scale, datapoint.y0, datapoint.y1, relAccessor);
}
