// @dynatrace/strato-components/charts/xy-chart/components/displays/RectDisplay.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var RectDisplay_exports = {};
__export(RectDisplay_exports, {
  RectDisplay: () => RectDisplay
});
module.exports = __toCommonJS(RectDisplay_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_rect_accessors = require("./rect-accessors.js");
var import_BarCanvas = require("../../../core/components/canvas/components/BarCanvas.js");
var import_useGridPlotSize = require("../../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_constants = require("../../../core/constants.js");
var import_safe_get = require("../../../core/utils/safe-get.js");
var import_defaults = require("../../constants/defaults.js");
var import_useXYChartColorScales = require("../../hooks/useXYChartColorScales.js");
var import_useXYChartData = require("../../hooks/useXYChartData.js");
var import_get_scale = require("../../utils/get-scale.js");
var import_validation_datapoint = require("../../utils/validation-datapoint.js");
var import_value_representation = require("../../utils/value-representation.js");
function RectDisplay(props) {
  const { axisIdToScales, series, rectFilterCallback } = props;
  const { height, width } = (0, import_useGridPlotSize.useGridPlotSize)();
  const colorScalesMap = (0, import_useXYChartColorScales.useXYChartColorScales)();
  const { datapointMetadataMap, axisIdToScaleDomain, seriesToMetadata } = (0, import_useXYChartData.useXYChartData)();
  return series.map((series2) => {
    const xScale = (0, import_get_scale.getCurrentScale)(axisIdToScales, series2.xAxisId);
    const yScale = (0, import_get_scale.getCurrentScale)(axisIdToScales, series2.yAxisId);
    const { xAxis, yAxis } = (0, import_safe_get.safeGet)(seriesToMetadata, series2);
    const xRelAccessor = (0, import_value_representation.relAccessor)(xAxis, axisIdToScaleDomain);
    const yRelAccessor = (0, import_value_representation.relAccessor)(yAxis, axisIdToScaleDomain);
    const colorScale = colorScalesMap.get(series2.id);
    return colorScale ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_BarCanvas.BarCanvas,
      {
        data: rectFilterCallback ? series2.filteredData.filter(rectFilterCallback) : series2.filteredData,
        canvasHeight: height,
        canvasWidth: width,
        color: (0, import_rect_accessors.colorAccessor)(colorScale, datapointMetadataMap),
        x: (0, import_rect_accessors.xAccessor)(xScale, xRelAccessor),
        width: (0, import_rect_accessors.widthAccessor)(xScale, xRelAccessor),
        y: (0, import_rect_accessors.yAccessor)(yScale, yRelAccessor),
        height: (0, import_rect_accessors.heightAccessor)(yScale, yRelAccessor),
        opacity: () => import_constants.DEFAULT_FILL_OPACITY,
        strokeWidth: import_defaults.RECT_STROKE_WIDTH,
        defined: (0, import_validation_datapoint.isGapDatapointForDisplay)(datapointMetadataMap)
      },
      series2.id
    ) : null;
  });
}
RectDisplay["displayName"] = "RectDisplay";
