// @dynatrace/strato-components/esm/core/sdk/sdk-logger.js
import { getAppId } from "@dynatrace-sdk/app-environment";
const getGlobalWithDtRuntime = () => typeof globalThis !== "undefined" ? globalThis : window;
const supportsLoggerV2 = () => typeof getGlobalWithDtRuntime().dtRuntime?.logger?.createLoggerV2 === "function";
const supportsLoggerV1 = () => typeof getGlobalWithDtRuntime().dtRuntime?.logger?.createLogger === "function";
const LOGGER_FACTORY = /* @__PURE__ */ (() => {
  let instance;
  const createLogger = () => {
    const dtRuntimeLogger = getGlobalWithDtRuntime().dtRuntime?.logger;
    if (dtRuntimeLogger === void 0) {
      return null;
    }
    if (supportsLoggerV2()) {
      return dtRuntimeLogger.createLoggerFactoryV2({
        context: {
          id: "@dynatrace/strato-design-system",
          // @ts-expect-error Global variable injected at package build time
          version: "3.6.0"
        }
      });
    } else if (supportsLoggerV1()) {
      return dtRuntimeLogger.createLoggerFactory({
        context: {
          id: "@dynatrace/strato-design-system",
          // @ts-expect-error Global variable injected at package build time
          version: "3.6.0"
        }
      });
    }
    return null;
  };
  return {
    getInstace: () => {
      if (!instance) {
        instance = createLogger();
      }
      return instance;
    }
  };
})();
const createSdkLogger = (name) => {
  if (LOGGER_FACTORY) {
    return LOGGER_FACTORY.getInstace()?.createLogger(name);
  } else {
    const wrapConsole = (logFunc) => (...args) => logFunc(`[${name}]: ${args[0]}`, args[1]);
    return Object.freeze({
      info: wrapConsole(console.log),
      warn: wrapConsole(console.warn),
      error: wrapConsole(console.error)
    });
  }
};
const isBuiltInAppid = (appId) => appId.startsWith("dynatrace.");
const isComponentDebugLoggingEnabled = (componentName) => window.localStorage.getItem(`_strato_${componentName}_logging_enabled_`) === "true";
function createStratoLogger(componentName) {
  const isBuiltInApp = isBuiltInAppid(getAppId());
  const logger = createSdkLogger("Strato: " + componentName);
  const log = (level, message, error2, details = {}) => {
    if (logger === void 0) {
      return;
    }
    const logMessage = message?.trim() || "An error occurred";
    const shouldLog = isBuiltInApp || isComponentDebugLoggingEnabled(componentName);
    if (!shouldLog) {
      return;
    }
    let context = "undefined";
    if (this) {
      try {
        context = JSON.stringify(this);
      } catch (e) {
        console.error(e);
        context = "could not be stringified";
      }
    }
    if (supportsLoggerV2()) {
      logger[level](`${logMessage}; context: ${context}`, {
        error: error2,
        details
      });
    } else {
      try {
        logger[level](
          `${logMessage}; details: ${JSON.stringify(
            details
          )}; context: ${context};`,
          error2
        );
      } catch (e) {
        console.error(e);
        logger[level](
          `${logMessage}; details: details could not be stringified; context: ${context};`,
          error2
        );
      }
    }
  };
  const error = (message, error2, details) => {
    log("error", message, error2, details);
  };
  const warn = (message, error2, details) => {
    log("warn", message, error2, details);
  };
  const info = (message, error2, details) => {
    log("info", message, error2, details);
  };
  return {
    error,
    warn,
    info
  };
}
const stratoLoggers = {};
const getStratoLogger = (componentName) => {
  if (!stratoLoggers[componentName]) {
    stratoLoggers[componentName] = createStratoLogger(componentName);
  }
  return stratoLoggers[componentName];
};
export {
  createStratoLogger,
  getGlobalWithDtRuntime,
  getStratoLogger,
  supportsLoggerV1,
  supportsLoggerV2
};
