// @dynatrace/strato-components/esm/forms/date-time-range-picker/hooks/useDateTimeRangePicker.js
import { TZDateMini } from "@date-fns/tz";
import {
  useCallback,
  useMemo,
  useRef
} from "react";
import { useIntl } from "react-intl";
import { parseTimeAsTimeValue } from "@dynatrace-sdk/units";
import { getTimezone } from "@dynatrace-sdk/user-preferences";
import { isFullTimeframe } from "../../../core/time/helpers.js";
import { useDateTimeRangePickerStore } from "../../../core/time/timeframeStore/useTimeframeStore.js";
import { focusDateTimeInput } from "../../date-time-picker/utils/focusDateTimeInput.js";
import {
  captureUndoSnapshot,
  isContentModifyingKey,
  normalizeISOInput,
  parseDateOnlyToISO,
  parseTimeOnlyToISO
} from "../helpers.js";
import {
  getExpressionCursorOffset,
  restoreExpressionCursorOffset
} from "../selection-helpers.js";
const restoreSpinButtons = (ref, spinButtonTexts) => {
  spinButtonTexts.forEach((text, type) => {
    const handler = ref.current?.spinButtonHandlers.get(type) ?? null;
    if (!handler) {
      return;
    }
    if (text === "") {
      handler.clearValue(true);
    } else if (handler.type === "number") {
      handler.applyText(text, true);
    } else {
      handler.searchValue(text[0] ?? "", true);
    }
  });
};
const restoreSnapshot = (snapshot, ref) => {
  if (snapshot.valueType === "iso8601") {
    restoreSpinButtons(ref, snapshot.spinButtonTexts);
  } else if (ref.current?.expressionInputRef) {
    const invalidExpresssion = !snapshot.value?.absoluteDate;
    if (invalidExpresssion) {
      restoreSpinButtons(ref, snapshot.spinButtonTexts);
    }
    ref.current.expressionInputRef.textContent = snapshot.expressionText;
  }
};
const useDateTimeRangePicker = (props) => {
  const { placeholder, readOnly, dateTimeFromRef, dateTimeToRef } = props;
  const { formatMessage } = useIntl();
  const value = useDateTimeRangePickerStore(
    (state) => state.dateTimeRangePickerTimeframe
  );
  const setValue = useDateTimeRangePickerStore(
    (state) => state.setDateTimeRangePickerTimeframe
  );
  const pasteUndoSnapshotRef = useRef(null);
  const localizedPlaceholder = useMemo(
    () => ({
      from: placeholder?.from ?? formatMessage({
        defaultMessage: "Start time",
        id: "gnKOHMjpyzk6hNcF",
        description: "Start of timeframe"
      }),
      to: placeholder?.to ?? formatMessage({
        defaultMessage: "End time",
        id: "sGxf5yRVhyC8Fpo/",
        description: "End of timeframe"
      })
    }),
    [formatMessage, placeholder?.from, placeholder?.to]
  );
  const onCopyCapture = useCallback(
    (event) => {
      event.preventDefault();
      event.stopPropagation();
      if (!isFullTimeframe(value)) {
        return;
      }
      navigator.clipboard.writeText(`${value.from.value}/${value.to.value}`).catch(() => {
      });
    },
    [value]
  );
  const onPasteCapture = useCallback(
    (e) => {
      if (readOnly) {
        return;
      }
      const data = e.clipboardData?.getData("text/plain");
      if (!data) {
        e.preventDefault();
        return;
      }
      if (!dateTimeFromRef.current || !dateTimeToRef.current) {
        return;
      }
      const focusTarget = document.activeElement;
      const isExpressionInputFocused = focusTarget === dateTimeFromRef.current?.expressionInputRef || focusTarget === dateTimeToRef.current?.expressionInputRef;
      const focusedExpression = isExpressionInputFocused ? focusTarget : null;
      const expressionCursorOffset = focusedExpression ? getExpressionCursorOffset(focusedExpression) : void 0;
      const fromValueType = dateTimeFromRef.current?.element?.getAttribute("data-valuetype") ?? "expression";
      const toValueType = dateTimeToRef.current?.element?.getAttribute("data-valuetype") ?? "expression";
      const fromSnapshot = captureUndoSnapshot(
        dateTimeFromRef.current.spinButtonHandlers,
        dateTimeFromRef.current.expressionInputRef,
        value?.from ?? null,
        fromValueType
      );
      const toSnapshot = captureUndoSnapshot(
        dateTimeToRef.current.spinButtonHandlers,
        dateTimeToRef.current.expressionInputRef,
        value?.to ?? null,
        toValueType
      );
      const trimmed = data.trim();
      const hasArrowSep = trimmed.includes("\u2192");
      const parts = hasArrowSep ? trimmed.split("\u2192") : trimmed.split("/");
      const [nextFrom, nextTo] = parts;
      if (nextFrom && nextTo) {
        if (hasArrowSep) {
          pasteUndoSnapshotRef.current = {
            from: fromSnapshot,
            to: toSnapshot,
            focusTarget,
            expressionCursorOffset
          };
          dateTimeFromRef.current?.pasteText(nextFrom, true);
          dateTimeToRef.current?.pasteText(nextTo, true);
          e.preventDefault();
          e.stopPropagation();
          requestAnimationFrame(() => {
            focusDateTimeInput(null, dateTimeFromRef, dateTimeToRef);
          });
          return;
        }
        if (parts.length === 2) {
          const relativeDate = TZDateMini.tz(getTimezone()).getTime();
          const parsedFrom = parseDateOnlyToISO(nextFrom) ?? parseTimeOnlyToISO(nextFrom) ?? normalizeISOInput(nextFrom);
          const parsedTo = parseDateOnlyToISO(nextTo, "endOfDay") ?? parseTimeOnlyToISO(nextTo) ?? normalizeISOInput(nextTo);
          const validFrom = parseTimeAsTimeValue(parsedFrom, relativeDate);
          const validTo = parseTimeAsTimeValue(parsedTo, relativeDate);
          if (validFrom && validTo) {
            pasteUndoSnapshotRef.current = {
              from: fromSnapshot,
              to: toSnapshot,
              focusTarget
            };
            dateTimeFromRef.current?.pasteText(nextFrom, true);
            dateTimeToRef.current?.pasteText(nextTo, true);
            e.preventDefault();
            e.stopPropagation();
            return;
          }
        }
      }
      pasteUndoSnapshotRef.current = {
        from: fromSnapshot,
        to: toSnapshot,
        focusTarget,
        expressionCursorOffset
      };
    },
    [readOnly, value, dateTimeFromRef, dateTimeToRef]
  );
  const onUndoPasteKeyDown = useCallback(
    (event) => {
      const isCtrlZ = (event.ctrlKey || event.metaKey) && event.key === "z";
      if (isCtrlZ && pasteUndoSnapshotRef.current) {
        const snapshot = pasteUndoSnapshotRef.current;
        pasteUndoSnapshotRef.current = null;
        event.preventDefault();
        event.stopPropagation();
        restoreSnapshot(snapshot.from, dateTimeFromRef);
        restoreSnapshot(snapshot.to, dateTimeToRef);
        const fromValue = snapshot.from.value ?? void 0;
        const toValue = snapshot.to.value ?? void 0;
        setValue(
          fromValue || toValue ? { from: fromValue, to: toValue } : null
        );
        if (snapshot.focusTarget instanceof HTMLElement && snapshot.focusTarget !== document.activeElement) {
          snapshot.focusTarget.focus();
        }
        if (snapshot.expressionCursorOffset !== void 0 && snapshot.focusTarget instanceof HTMLSpanElement) {
          restoreExpressionCursorOffset(
            snapshot.focusTarget,
            snapshot.expressionCursorOffset
          );
        }
      } else if (pasteUndoSnapshotRef.current && isContentModifyingKey(event)) {
        pasteUndoSnapshotRef.current = null;
      }
    },
    [dateTimeFromRef, dateTimeToRef, setValue]
  );
  return {
    placeholder: localizedPlaceholder,
    onCopyCapture,
    onPasteCapture,
    onUndoPasteKeyDown
  };
};
export {
  useDateTimeRangePicker
};
