// @dynatrace/strato-components/charts/gantt-chart/components/GanttSegment.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var GanttSegment_exports = {};
__export(GanttSegment_exports, {
  GanttSegment: () => GanttSegment
});
module.exports = __toCommonJS(GanttSegment_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_defaults = require("../constants/defaults.js");
var import_GanttSegmentsContext = require("../context/GanttSegmentsContext.js");
var import_get_entry_color = require("../functions/get-entry-color.js");
var import_use_gantt_entry_handlers = require("../hooks/use-gantt-entry-handlers.js");
var import_use_gantt_segment_position = require("../hooks/use-gantt-segment-position.js");
const GanttSegment = (props) => {
  const { start, end, height, containerWidth, data } = props;
  const {
    segmentsData,
    parentSegmentsData,
    config: { colorAccessor, colorPalette }
  } = (0, import_react.useContext)(import_GanttSegmentsContext.GanttSegmentsContext);
  const fill = data ? (0, import_get_entry_color.getEntryColor)(data, colorAccessor, colorPalette) : `url(#${import_defaults.STRIPES_PATTERN_ID})`;
  const dataForEntryHandlers = (0, import_react.useMemo)(
    () => data ? {
      segment: data,
      row: segmentsData,
      parent: parentSegmentsData
    } : void 0,
    [data, segmentsData, parentSegmentsData]
  );
  const {
    onMouseClickHandler,
    onMouseEnterHandler,
    onMouseLeaveHandler,
    onMouseMoveHandler
  } = (0, import_use_gantt_entry_handlers.useGanttEntryHandlers)(dataForEntryHandlers);
  if (!isFinite(start) || containerWidth < 0 || end !== void 0 && !isFinite(end)) {
    throw new Error("Invalid input data");
  }
  const {
    start: segmentXStart,
    end: segmentXEnd,
    width: segmentWidth,
    isNegative
  } = (0, import_use_gantt_segment_position.useGanttSegmentPosition)(start, end ?? start);
  const isTimestamp = end === void 0 || start === end || segmentWidth < import_defaults.TIMESTAMP_WIDTH;
  const segmentX = isNegative ? segmentXEnd : segmentXStart;
  const x = isTimestamp ? segmentX - 1 : segmentX;
  const width = isTimestamp ? import_defaults.TIMESTAMP_WIDTH : segmentWidth;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "rect",
    {
      onMouseEnter: onMouseEnterHandler,
      onClick: onMouseClickHandler,
      onMouseLeave: onMouseLeaveHandler,
      onMouseMove: onMouseMoveHandler,
      x,
      y: 0,
      opacity: import_defaults.DEFAULT_BAR_OPACITY,
      width,
      height,
      fill,
      pointerEvents: "auto"
    }
  );
};
