// @dynatrace/strato-components/charts/gantt-chart/components/GanttChartLineRenderer.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var GanttChartLineRenderer_exports = {};
__export(GanttChartLineRenderer_exports, {
  GanttChartLineRenderer: () => GanttChartLineRenderer
});
module.exports = __toCommonJS(GanttChartLineRenderer_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_group = require("@visx/group");
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_GanttChartLineRenderer_css = require("./GanttChartLineRenderer.sty.js");
var import_GanttSegment = require("./GanttSegment.js");
var import_ChartStateHandler = require("../../core/components/chart-states/ChartStateHandler.js");
var import_get_chart_states_template = require("../../core/components/chart-states/utils/get-chart-states-template.js");
var import_EmptyStateSlot = require("../../core/slots/chart-states/EmptyStateSlot.js");
var import_ErrorStateSlot = require("../../core/slots/chart-states/ErrorStateSlot.js");
var import_defaults = require("../constants/defaults.js");
var import_GanttSegmentsContext = require("../context/GanttSegmentsContext.js");
const _GanttChartLineRenderer = (0, import_react.forwardRef)((props, forwardedRef) => {
  const { width, height: svgHeight, children } = props;
  const { segmentsData } = (0, import_GanttSegmentsContext.useGanttSegmentsContext)();
  const backgroundSegments = [];
  const { emptyState, errorState } = (0, import_get_chart_states_template.getChartStatesTemplates)(children);
  if (props.showBackground) {
    segmentsData.sort((a, b) => a.start - b.start);
    for (let i = 0; i < segmentsData.length - 1; i++) {
      const currentSegmentEnd = segmentsData[i].end || segmentsData[i].start;
      const nextSegmentStart = segmentsData[i + 1].start;
      if (currentSegmentEnd < nextSegmentStart) {
        backgroundSegments.push({
          start: currentSegmentEnd,
          end: nextSegmentStart
        });
      }
    }
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_ChartStateHandler.ChartStateHandler,
    {
      empty: (0, import_lodash_es.isEmpty)(segmentsData),
      height: "100%",
      emptyState,
      errorState,
      isLoading: props.loading === true,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: import_GanttChartLineRenderer_css.svgWrapperCSS, ref: forwardedRef, children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("svg", { width, height: svgHeight, children: [
        backgroundSegments.length > 0 && /* @__PURE__ */ (0, import_jsx_runtime.jsx)("defs", { children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
          "pattern",
          {
            id: import_defaults.STRIPES_PATTERN_ID,
            patternUnits: "userSpaceOnUse",
            width: import_defaults.DISABLED_PATTERN_SIZE * 2,
            height: import_defaults.DISABLED_PATTERN_SIZE * 2,
            patternTransform: `rotate(${import_defaults.DISABLED_PATTERN_ANGLE})`,
            children: [
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                "rect",
                {
                  x: "0",
                  y: "0",
                  width: import_defaults.DISABLED_PATTERN_SIZE,
                  height: import_defaults.DISABLED_PATTERN_SIZE * 2,
                  fill: import_defaults.DEFAULT_GANTT_COLORS.background
                }
              ),
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                "rect",
                {
                  x: import_defaults.DISABLED_PATTERN_SIZE,
                  y: "0",
                  width: import_defaults.DISABLED_PATTERN_SIZE,
                  height: import_defaults.DISABLED_PATTERN_SIZE * 2,
                  fill: "transparent"
                }
              )
            ]
          }
        ) }),
        backgroundSegments.map((segment, index) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_GanttSegment.GanttSegment,
          {
            containerWidth: width,
            start: segment.start,
            end: segment.end,
            height: svgHeight
          },
          index
        )),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_group.Group, { children: segmentsData.map((entry, index) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_GanttSegment.GanttSegment,
          {
            containerWidth: width,
            height: svgHeight,
            data: entry,
            start: entry.start,
            end: entry.end
          },
          index
        )) })
      ] }) })
    }
  );
});
const GanttChartLineRenderer = Object.assign(_GanttChartLineRenderer, {
  EmptyState: import_EmptyStateSlot.EmptyStateSlot,
  ErrorState: import_ErrorStateSlot.ErrorStateSlot
});
