// @dynatrace/strato-components/esm/tables/DataTable/features/Config.js
import { useImperativeHandle, useRef } from "react";
import { useMergeRefs } from "../../../core/hooks/useMergeRefs.js";
import { useDataTableConfigElement } from "../../hooks/useDataTableConfig.js";
import { DATA_TABLE_DETAILS_ROW_IDENTIFIER } from "../constants.js";
function useDataTableConfigRef(forwardedRef, columnSettingsModalRef, table, props) {
  const containerRef = useRef(null);
  const rootRef = useMergeRefs([
    forwardedRef,
    containerRef
  ]);
  const configRowThresholds = useDataTableConfigElement("rowThresholds");
  useImperativeHandle(forwardedRef, () => ({
    element: containerRef.current,
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    getConfig() {
      const options = table.options;
      const state = options.state;
      return {
        pagination: {
          defaultPageIndex: state.pagination?.pageIndex,
          defaultPageSize: state.pagination?.pageSize
        },
        columnVisibility: state.columnVisibility,
        fullWidth: options.layout?.fullWidth,
        fullHeight: options.layout?.fullHeight,
        lineWrap: state.columnLineWrap,
        resizable: options.enableColumnResizing,
        selectableRows: options.enableRowSelection,
        // NOTE: typecast necessary due to generic resolving to unknown
        rowThresholds: props.rowThresholds ?? configRowThresholds,
        sortable: options.enableSorting,
        variant: state.visualVariantConfig
      };
    },
    downloadData(subset, excludeColumns) {
      switch (subset) {
        case "all":
          table.downloadAllData(excludeColumns);
          break;
        case "page":
          table.downloadPageData(excludeColumns);
          break;
        case "selected":
          table.downloadSelectedData(excludeColumns);
          break;
      }
    },
    scrollToRow: table.scrollToRow,
    getDisplayedRowIds() {
      const result = [];
      for (const row of table.getRowModel().rows) {
        if (!row[DATA_TABLE_DETAILS_ROW_IDENTIFIER]) {
          result.push(row.id);
        }
      }
      return result;
    },
    openColumnSettings: (columnSettings) => {
      columnSettingsModalRef.current?.openColumnSettings(columnSettings);
    }
  }));
  return rootRef;
}
export {
  useDataTableConfigRef
};
