// @dynatrace/strato-components/charts/core/components/legend/range/utils/labels-collision.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var labels_collision_exports = {};
__export(labels_collision_exports, {
  calculateOptimalEllipsis: () => calculateOptimalEllipsis,
  calculateOptimalPlacement: () => calculateOptimalPlacement,
  getLabelsPosition: () => getLabelsPosition,
  truncateLabel: () => truncateLabel
});
module.exports = __toCommonJS(labels_collision_exports);
var import_constants = require("../constants.js");
var import_get_horizontal_label_position = require("./get-horizontal-label-position.js");
var import_text_truncation = require("../../../../utils/text-truncation.js");
var import_chars_font_width_map = require("../../../text-ellipsis/chars-font-width-map.js");
var import_truncate_string = require("../../../text-ellipsis/truncate-string.js");
const calculateOptimalPlacement = (labelStart, labelEnd, startBound, endBound) => {
  if (labelStart < startBound) {
    const gap = startBound - labelStart;
    return labelStart + gap;
  }
  if (labelEnd > endBound) {
    const gap = endBound - labelEnd;
    return labelStart + gap;
  }
  return labelStart;
};
const calculateOptimalEllipsis = (text, startPosition, endPosition, truncationMode, startBound, endBound) => {
  let candidateStartPosX = startPosition;
  let candidateEndPosX = endPosition;
  let candidateLabel = text;
  let iteration = 0;
  do {
    candidateLabel = truncateLabel(candidateLabel, truncationMode);
    const newLabelWidth = (0, import_truncate_string.calculateWidth)(
      candidateLabel,
      import_chars_font_width_map.charsFontWidth12pixelsSize500Map
    );
    const { x: newCandidateStartPosX } = (0, import_get_horizontal_label_position.getHorizontalLabelPosition)(
      newLabelWidth,
      startBound,
      startBound + newLabelWidth,
      "start"
    );
    candidateStartPosX = newCandidateStartPosX;
    candidateEndPosX = candidateStartPosX + newLabelWidth;
    iteration++;
    if (iteration > text.length) {
      break;
    }
  } while (candidateStartPosX < startBound || candidateEndPosX > endBound);
  return { text: candidateLabel, x: candidateStartPosX };
};
const truncateLabel = (text, truncationMode) => {
  const filteredText = (text ?? "").replace(import_constants.TEXT_ELLIPSIS, "");
  if (filteredText.length < 2) {
    return text;
  }
  return import_text_truncation.TRUNCATE_LABEL[truncationMode](filteredText);
};
const getLabelsPosition = (startLabel, endLabel, bounds) => {
  const startLabelWidth = (0, import_truncate_string.calculateWidth)(
    startLabel,
    import_chars_font_width_map.charsFontWidth12pixelsSize500Map
  );
  const endLabelWidth = (0, import_truncate_string.calculateWidth)(
    endLabel,
    import_chars_font_width_map.charsFontWidth12pixelsSize500Map
  );
  return {
    startLabel: { start: bounds[0], end: bounds[0] + startLabelWidth },
    endLabel: { start: bounds[1] - endLabelWidth, end: bounds[1] }
  };
};
