// @dynatrace/strato-components/charts/core/components/legend/range/components/RangeLegendBoundaryLabel/RangeLegendBoundaryLabel.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var RangeLegendBoundaryLabel_exports = {};
__export(RangeLegendBoundaryLabel_exports, {
  RangeLegendBoundaryLabel: () => RangeLegendBoundaryLabel
});
module.exports = __toCommonJS(RangeLegendBoundaryLabel_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_RangeLegendBoundaryLabel_css = require("./RangeLegendBoundaryLabel.sty.js");
var import_constants = require("../../../../../constants.js");
var import_chars_font_width_map = require("../../../../text-ellipsis/chars-font-width-map.js");
var import_truncate_string = require("../../../../text-ellipsis/truncate-string.js");
var import_constants2 = require("../../constants.js");
var import_RangeLegendOrientationContext = require("../../context/RangeLegendOrientationContext.js");
var import_selectors = require("../../hooks/store/selectors.js");
var import_get_horizontal_label_position = require("../../utils/get-horizontal-label-position.js");
var import_get_vertical_label_position = require("../../utils/get-vertical-label-position.js");
const RangeLegendBoundaryLabel = (props) => {
  const { alignment, endPosition, startPosition, text, ...textProps } = props;
  const { orientation } = (0, import_RangeLegendOrientationContext.useOrientation)();
  const hoveredMarkerId = (0, import_selectors.useRangeLegendHoveredMarkerId)();
  const labelTextWidth = (0, import_truncate_string.calculateWidth)(text, import_chars_font_width_map.charsFontWidth12pixelsSize500Map);
  const labelTextHeight = import_truncate_string.TEXT_HEIGHT;
  const isVertical = orientation === import_constants.ORIENTATION.Vertical;
  const { x: labelX, y: labelY } = !isVertical ? (0, import_get_horizontal_label_position.getHorizontalLabelPosition)(
    labelTextWidth,
    startPosition,
    endPosition,
    alignment
  ) : (0, import_get_vertical_label_position.getVerticalLabelPosition)(
    labelTextHeight,
    startPosition,
    endPosition,
    alignment
  );
  const isMarkerShown = hoveredMarkerId !== null;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "text",
    {
      x: labelX + (isVertical ? import_constants2.LABEL_GAP : 0),
      y: labelY + (!isVertical ? import_constants2.LABEL_GAP : 0),
      fontSize: import_constants2.DEFAULT_LABEL_FONT_SIZE,
      fontWeight: import_constants2.LABEL_FONT_WEIGHT,
      fill: import_constants2.DEFAULT_LABEL_FONT_COLOR,
      className: (0, import_RangeLegendBoundaryLabel_css.boundaryLabel)({ isMarkerShown }),
      role: "listitem",
      "aria-label": `Boundary ${text}`,
      ...textProps,
      children: text
    }
  );
};
