// @dynatrace/strato-components/esm/filters/timeframe-selector/components/TimeframeSelectorTooltip.js
import { jsx, jsxs } from "react/jsx-runtime";
import { useCallback, useState } from "react";
import { Flex } from "../../../layouts/flex/Flex.js";
import { Tooltip } from "../../../overlays/tooltip/Tooltip.js";
import { Text } from "../../../typography/text/Text.js";
import { useTooltipMessages } from "../hooks/useTooltipMessages.js";
const TimeframeSelectorTooltip = (props) => {
  const {
    value,
    disabled,
    displayValue,
    children
  } = props;
  const [tooltipTimeValues, setTooltipTimeValues] = useState({
    tooltipFormattedTimeValueStart: "",
    tooltipFormattedTimeValueEnd: ""
  });
  const {
    tooltipFormattedTimeValueStart,
    tooltipFormattedTimeValueEnd,
    timeframeDuration,
    isAbsoluteTimeframe,
    isInvalid
  } = useTooltipMessages(value);
  const onOpenTooltip = useCallback(
    (open) => {
      if (open) {
        setTooltipTimeValues({
          tooltipFormattedTimeValueStart,
          tooltipFormattedTimeValueEnd
        });
      }
    },
    [tooltipFormattedTimeValueEnd, tooltipFormattedTimeValueStart]
  );
  return /* @__PURE__ */ jsx(
    Tooltip,
    {
      text: /* @__PURE__ */ jsxs(Flex, { flexDirection: "column", gap: 4, children: [
        /* @__PURE__ */ jsxs(Flex, {
          gap: 0,
          flexDirection: "column",
          // displayValue is a string when a preset item name is used.
          children: [
            (typeof displayValue === "string" || !isAbsoluteTimeframe) && /* @__PURE__ */ jsx(Text, { textStyle: "small-emphasized", children: displayValue }),
            /* @__PURE__ */ jsx(Text, { textStyle: "small", children: tooltipTimeValues.tooltipFormattedTimeValueStart }),
            /* @__PURE__ */ jsx(Text, { textStyle: "small", children: tooltipTimeValues.tooltipFormattedTimeValueEnd })
          ]
        }),
        /* @__PURE__ */ jsx(Flex, { gap: 0, flexDirection: "column", children: /* @__PURE__ */ jsx(Text, { textStyle: "small", children: timeframeDuration }) })
      ] }),
      disabled: disabled || isInvalid,
      onOpenChange: onOpenTooltip,
      children
    }
  );
};
TimeframeSelectorTooltip.displayName = "TimeframeSelectorTooltip";
export {
  TimeframeSelectorTooltip
};
