// @dynatrace/strato-components/charts/timeseries/constants.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var constants_exports = {};
__export(constants_exports, {
  AREA_CHART_DISPLAY_ARIA_LABEL: () => AREA_CHART_DISPLAY_ARIA_LABEL,
  BAR_CHART_DISPLAY_ARIA_LABEL: () => BAR_CHART_DISPLAY_ARIA_LABEL,
  DEFAULT_SCALE_FALLBACK_COLOR: () => DEFAULT_SCALE_FALLBACK_COLOR,
  DISPLAY_NAME: () => DISPLAY_NAME,
  LINE_CHART_DISPLAY_ARIA_LABEL: () => LINE_CHART_DISPLAY_ARIA_LABEL,
  TIMESERIES_ARIA_LABEL: () => TIMESERIES_ARIA_LABEL,
  TIMESERIES_CHART_DEFAULT_VARIANT: () => TIMESERIES_CHART_DEFAULT_VARIANT,
  TIMESERIES_INTERACTIONS_ARIA_LABEL: () => TIMESERIES_INTERACTIONS_ARIA_LABEL,
  TIMESERIES_TIME_AXIS_ARIA_LABEL: () => TIMESERIES_TIME_AXIS_ARIA_LABEL,
  defaultColorPalette: () => defaultColorPalette,
  defaultGapPolicy: () => defaultGapPolicy,
  defaultInfiniteZoom: () => defaultInfiniteZoom,
  defaultInteractions: () => defaultInteractions,
  defaultLegend: () => defaultLegend,
  defaultPointsDisplay: () => defaultPointsDisplay,
  defaultSeriesCurve: () => defaultSeriesCurve,
  defaultSeriesDisplayMode: () => defaultSeriesDisplayMode,
  defaultThreshold: () => defaultThreshold,
  defaultTimeseriesChartConfig: () => defaultTimeseriesChartConfig,
  defaultTimeseriesPerVariant: () => defaultTimeseriesPerVariant,
  defaultToolbar: () => defaultToolbar,
  defaultTooltipVariant: () => defaultTooltipVariant,
  defaultValueRepresentation: () => defaultValueRepresentation,
  defaultYAxis: () => defaultYAxis,
  initialState: () => initialState,
  zIndicesTimeseriesAreas: () => zIndicesTimeseriesAreas
});
module.exports = __toCommonJS(constants_exports);
var import_AnnotationsDataInitialState = require("../core/components/annotations/providers/AnnotationsDataInitialState.js");
var import_constants = require("../core/constants.js");
var import_gap_policy = require("../core/types/gap-policy.js");
const DISPLAY_NAME = "TimeseriesChart";
const defaultTooltipVariant = "single";
const defaultSeriesDisplayMode = "multi-line";
const defaultYAxis = {};
const defaultGapPolicy = import_gap_policy.GAP_POLICY_GAP;
const defaultSeriesCurve = "linear";
const defaultPointsDisplay = "auto";
const defaultValueRepresentation = "absolute";
const defaultInfiniteZoom = false;
const DEFAULT_SCALE_FALLBACK_COLOR = import_constants.SCALE_FALLBACK_COLOR;
const defaultLegend = {
  hidden: false,
  position: "auto",
  ratio: import_constants.defaultRatio
};
const defaultToolbar = {
  placement: "top-right",
  enabled: true
};
const defaultInteractions = {
  onZoomChange: void 0,
  pan: {
    enabled: true
  },
  zoom: {
    enabled: true
  }
};
const defaultColorPalette = "categorical";
const defaultTimeseriesChartConfig = {
  legend: defaultLegend,
  colorPalette: defaultColorPalette,
  gapPolicy: defaultGapPolicy,
  pointsDisplay: defaultPointsDisplay,
  valueRepresentation: defaultValueRepresentation,
  tooltip: {
    variant: defaultTooltipVariant,
    seriesDisplayMode: defaultSeriesDisplayMode
  },
  toolbar: defaultToolbar,
  interactions: defaultInteractions
};
const defaultTimeseriesPerVariant = {
  line: [],
  area: [],
  bar: [],
  band: [],
  shapePriority: []
};
const defaultThreshold = {
  left: [],
  right: []
};
const TIMESERIES_CHART_DEFAULT_VARIANT = "line";
const zIndicesTimeseriesAreas = {
  thresholdRange: 0,
  plot: 1,
  thresholdLine: 2,
  overlays: 3
};
const initialState = {
  filteredSeriesIds: /* @__PURE__ */ new Set(),
  highlightedSeriesId: void 0,
  annotations: import_AnnotationsDataInitialState.initialSelectionState,
  zoom: {
    zoomInEnabled: true,
    zoomOutEnabled: false
  },
  currentDomain: void 0,
  select: void 0,
  interactionMode: "explore",
  explore: {
    visible: false,
    pinned: false,
    position: void 0
  },
  thresholds: { hoveredThresholdId: void 0 }
};
const TIMESERIES_ARIA_LABEL = "Timeseries Chart";
const TIMESERIES_INTERACTIONS_ARIA_LABEL = "Timeseries Chart interactions";
const TIMESERIES_TIME_AXIS_ARIA_LABEL = "timeseries-time-axis";
const AREA_CHART_DISPLAY_ARIA_LABEL = "area-chart-display";
const LINE_CHART_DISPLAY_ARIA_LABEL = "line-chart-display";
const BAR_CHART_DISPLAY_ARIA_LABEL = "bar-chart-display";
