// @dynatrace/strato-components/charts/timeseries/TimeseriesChartRoot.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TimeseriesChartRoot_exports = {};
__export(TimeseriesChartRoot_exports, {
  TimeseriesChartRoot: () => TimeseriesChartRoot
});
module.exports = __toCommonJS(TimeseriesChartRoot_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_TimeseriesChartLegend = require("./components/legend/TimeseriesChartLegend.js");
var import_constants = require("./constants.js");
var import_TimeseriesRawData_context = require("./context/TimeseriesRawData.context.js");
var import_useTimeseriesChartConfig = require("./hooks/useTimeseriesChartConfig.js");
var import_StoreProvider = require("./providers/StoreProvider.js");
var import_TimeseriesChartIdProvider = require("./providers/TimeseriesChartIdProvider.js");
var import_TimeseriesConfigProviders = require("./providers/TimeseriesConfigProviders.js");
var import_TimeseriesDataProvider = require("./providers/TimeseriesDataProvider.js");
var import_TimeseriesFilteredDataProvider = require("./providers/TimeseriesFilteredDataProvider.js");
var import_TimeseriesMinMaxXAxisProvider = require("./providers/TimeseriesMinMaxXAxisProvider.js");
var import_TimeseriesChartInternal = require("./TimeseriesChartInternal.js");
var import_timeseries = require("./utils/timeseries.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_ChartLayout = require("../core/components/chart-layout/ChartLayout.js");
var import_useAutoLegendRefresh = require("../core/components/chart-layout/hooks/useAutoLegendRefresh.js");
var import_get_chart_states_template = require("../core/components/chart-states/utils/get-chart-states-template.js");
var import_OverlayContext = require("../core/components/overlay-tooltip/contexts/OverlayContext.js");
var import_TooltipContext = require("../core/components/overlay-tooltip/contexts/TooltipContext.js");
var import_ChartSeriesActionContext = require("../core/components/series-action/ChartSeriesActionContext.js");
var import_coerceSizeValue = require("../core/utils/coerceSizeValue.js");
const TIMESERIES_CHART_DEFAULT_HEIGHT = 300;
const COMPONENT_NAME_BEHAVIOURAL_TRACKING = "TimeseriesChart";
const TimeseriesChartRoot = (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    children,
    data,
    gapPolicy,
    pointsDisplay,
    colorPalette,
    truncationMode,
    variant = import_constants.TIMESERIES_CHART_DEFAULT_VARIANT,
    valueRepresentation,
    height: heightProp,
    width,
    seriesActions,
    onMessage,
    loading = false,
    curve,
    initialZoom,
    currentZoom,
    onZoomChange,
    infiniteZoom,
    className,
    style,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow
  } = props;
  const [behaviorTrackingProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(props);
  const chartHeight = (0, import_coerceSizeValue.coerceSizeValue)(heightProp) || TIMESERIES_CHART_DEFAULT_HEIGHT;
  const dataPerVariant = (0, import_timeseries.splitTimeseriesPerVariant)(
    data ?? [],
    children,
    variant
  );
  const config = (0, import_useTimeseriesChartConfig.useTimeseriesChartConfig)(children, {
    gapPolicy,
    pointsDisplay,
    valueRepresentation,
    colorPalette,
    truncationMode,
    curve,
    initialZoom,
    currentZoom,
    onZoomChange,
    infiniteZoom
  });
  const { emptyState, errorState } = (0, import_get_chart_states_template.getChartStatesTemplates)(children);
  const { shapePriority, ...rawDataPerVariant } = dataPerVariant;
  const hasSeries = (0, import_timeseries.hasSeriesWithDatapoints)(dataPerVariant);
  const {
    position: legendPosition,
    onRatioChange,
    // eslint-disable-next-line @typescript-eslint/no-deprecated
    ratio,
    initialRatio,
    hidden: isLegendHidden
  } = config.legend;
  const legendRatio = initialRatio ?? ratio;
  const chartLayoutRef = (0, import_useAutoLegendRefresh.useAutoLegendRefresh)(data);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TimeseriesChartIdProvider.TimeseriesChartIdProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TimeseriesConfigProviders.TimeseriesConfigProvider, { config, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TimeseriesRawData_context.TimeseriesRawDataContext.Provider, { value: rawDataPerVariant, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_TimeseriesDataProvider.TimeseriesDataProvider,
    {
      data: dataPerVariant,
      valueRepresentation: config.valueRepresentation,
      onMessage,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TimeseriesMinMaxXAxisProvider.TimeseriesMinMaxXAxisProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_ChartSeriesActionContext.ChartSeriesActionContext.Provider,
        {
          value: {
            chartSeriesActions: seriesActions
          },
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_StoreProvider.StoreProvider,
            {
              data: dataPerVariant,
              currentDomain: config.currentZoom,
              initialDomain: config.initialZoom,
              infiniteZoom: config.infiniteZoom,
              children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TimeseriesFilteredDataProvider.TimeseriesFilteredDataProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayContext.OverlayProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TooltipContext.TooltipProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
                import_ChartLayout.ChartLayout,
                {
                  ref: chartLayoutRef,
                  chartHeight,
                  chartWidth: width,
                  isEmpty: !hasSeries,
                  emptyState,
                  errorState,
                  showLoader: loading,
                  className,
                  style,
                  ...behaviorTrackingProps,
                  "data-dtrum-mask": dataDtrumMask,
                  "data-dtrum-allow": dataDtrumAllow,
                  "data-dt-component": COMPONENT_NAME_BEHAVIOURAL_TRACKING,
                  children: [
                    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ChartLayout.ChartLayout.Graph, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                      import_TimeseriesChartInternal.TimeseriesChartInternal,
                      {
                        ref: forwardedRef,
                        config
                      }
                    ) }),
                    !isLegendHidden && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                      import_ChartLayout.ChartLayout.Legend,
                      {
                        position: legendPosition,
                        initialRatio: legendRatio,
                        onResize: onRatioChange,
                        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TimeseriesChartLegend.TimeseriesChartLegend, {})
                      }
                    )
                  ]
                }
              ) }) }) })
            }
          )
        }
      ) })
    }
  ) }) }) });
});
TimeseriesChartRoot.displayName = "TimeseriesChartRoot";
