// @dynatrace/strato-components/charts/timeseries/TimeseriesChart.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TimeseriesChart_exports = {};
__export(TimeseriesChart_exports, {
  TimeseriesAnnotations: () => TimeseriesAnnotations,
  TimeseriesChart: () => TimeseriesChart,
  TimeseriesChartSwitcher: () => TimeseriesChartSwitcher
});
module.exports = __toCommonJS(TimeseriesChart_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_Annotations = require("./slots/annotations/Annotations.js");
var import_Marker = require("./slots/annotations/Marker.js");
var import_Tooltip = require("./slots/annotations/Tooltip.js");
var import_Track = require("./slots/annotations/Track.js");
var import_DownloadCSV = require("./slots/DownloadCSV.js");
var import_Legend = require("./slots/Legend.js");
var import_Pan = require("./slots/Pan.js");
var import_Select = require("./slots/Select.js");
var import_Area = require("./slots/shapes/Area.js");
var import_Band = require("./slots/shapes/Band.js");
var import_Bar = require("./slots/shapes/Bar.js");
var import_Line = require("./slots/shapes/Line.js");
var import_ThresholdIndicator = require("./slots/ThresholdIndicator.js");
var import_Toolbar = require("./slots/Toolbar.js");
var import_Tooltip2 = require("./slots/Tooltip.js");
var import_XAxis = require("./slots/XAxis.js");
var import_YAxis = require("./slots/YAxis.js");
var import_Zoom = require("./slots/Zoom.js");
var import_TimeseriesChartRoot = require("./TimeseriesChartRoot.js");
var import_Intent = require("../../core/slots/Intent/Intent.js");
var import_EmptyStateSlot = require("../core/slots/chart-states/EmptyStateSlot.js");
var import_ErrorStateSlot = require("../core/slots/chart-states/ErrorStateSlot.js");
var import_ColorRule = require("../core/slots/ColorRule.js");
var import_is_feature_enabled = require("../core/utils/is-feature-enabled.js");
var import_TimeseriesToXY = require("../xy-chart-adoption/timeseries/TimeseriesToXY.js");
const TimeseriesChartSwitcher = (0, import_react.forwardRef)((props, forwardedRef) => {
  const { __useXYAdoption, ...rest } = props;
  if (__useXYAdoption ?? (0, import_is_feature_enabled.isFeatureEnabled)("timeseriesToXY")) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TimeseriesToXY.TimeseriesToXY, { ...rest, ref: forwardedRef });
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TimeseriesChartRoot.TimeseriesChartRoot, { ...rest, ref: forwardedRef });
});
TimeseriesChartSwitcher.displayName = "TimeseriesChart";
const TimeseriesChart = Object.assign(TimeseriesChartSwitcher, {
  Band: import_Band.Band,
  Legend: import_Legend.Legend,
  YAxis: import_YAxis.YAxis,
  XAxis: import_XAxis.XAxis,
  Tooltip: import_Tooltip2.Tooltip,
  Annotations: import_Annotations.Annotations,
  Line: import_Line.Line,
  Area: import_Area.Area,
  Bar: import_Bar.Bar,
  ThresholdIndicator: import_ThresholdIndicator.ThresholdIndicator,
  Select: import_Select.Select,
  Intent: import_Intent.Intent,
  EmptyState: import_EmptyStateSlot.EmptyStateSlot,
  ErrorState: import_ErrorStateSlot.ErrorStateSlot,
  ColorRule: import_ColorRule.ColorRule,
  Toolbar: import_Toolbar.Toolbar,
  DownloadCSV: import_DownloadCSV.TimeseriesChartDownloadCSVSlot,
  Pan: import_Pan.Pan,
  Zoom: import_Zoom.Zoom
});
const TimeseriesAnnotations = Object.assign(
  {},
  {
    Marker: import_Marker.Marker,
    Track: import_Track.Track,
    Tooltip: import_Tooltip.AnnotationsTooltip
  }
);
