// @dynatrace/strato-components/charts/timeseries/TimeseriesChartInternal.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TimeseriesChartInternal_exports = {};
__export(TimeseriesChartInternal_exports, {
  TimeseriesChartInternal: () => TimeseriesChartInternal
});
module.exports = __toCommonJS(TimeseriesChartInternal_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_constants = require("./components/annotations-timeseries/constants.js");
var import_TimeseriesAnnotationsIndicators = require("./components/annotations-timeseries/TimeseriesAnnotationsIndicators.js");
var import_TimeseriesChartAnnotationsArea = require("./components/annotations-timeseries/TimeseriesChartAnnotationsArea.js");
var import_TimeseriesTimeBottomAxisArea = require("./components/axes/TimeseriesTimeBottomAxisArea/TimeseriesTimeBottomAxisArea.js");
var import_TimeseriesValueLeftAxisArea = require("./components/axes/TimeseriesValueLeftAxisArea.js");
var import_TimeseriesValueRightAxisArea = require("./components/axes/TimeseriesValueRightAxisArea.js");
var import_TimeseriesChartOverlays = require("./components/overlays/TimeseriesChartOverlays.js");
var import_ThresholdIndicators = require("./components/thresholds/ThresholdIndicators.js");
var import_TimeseriesThresholdArea = require("./components/thresholds/TimeseriesThresholdArea.js");
var import_constants2 = require("./constants.js");
var import_useAnnotationsConfig = require("./hooks/useAnnotationsConfig.js");
var import_useChartInteractions = require("./hooks/useChartInteractions.js");
var import_useTimeseriesFilteredData = require("./hooks/useTimeseriesFilteredData.js");
var import_useTimeseriesFilteredYAxis = require("./hooks/useTimeseriesFilteredYAxis.js");
var import_useTimeseriesLegend = require("./hooks/useTimeseriesLegend.js");
var import_TimeseriesChartPlot = require("./layout/TimeseriesChartPlot.js");
var import_TimeseriesChartImperativeHandler = require("./providers/TimeseriesChartImperativeHandler.js");
var import_TimeseriesChartProviders = require("./providers/TimeseriesChartProviders.js");
var import_timeseries_grid_area_rules = require("./types/timeseries-grid-area-rules.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_defaultChartValues_css = require("../../core/styles/defaultChartValues.sty.js");
var import_constants3 = require("../core/components/annotations/constants.js");
var import_get_annotations_visible_area = require("../core/components/annotations/utils/get-annotations-visible-area.js");
var import_ChartGridRoot = require("../core/components/chart-grid/ChartGridRoot.js");
var import_useChartLayoutLoading = require("../core/components/chart-layout/hooks/useChartLayoutLoading.js");
var import_useChartLayoutOrientation = require("../core/components/chart-layout/hooks/useChartLayoutOrientation.js");
var import_useGraphSize = require("../core/components/chart-layout/hooks/useGraphSize.js");
var import_ChartLoaderOverlay = require("../core/components/chart-loader-overlay/ChartLoaderOverlay.js");
var import_constants4 = require("../core/components/data-point/constants.js");
const timeseriesGridTemplateAreas = `
  "${import_timeseries_grid_area_rules.AREAS.annotations} ${import_timeseries_grid_area_rules.AREAS.annotations} ${import_timeseries_grid_area_rules.AREAS.annotations} . ."
  "${import_timeseries_grid_area_rules.AREAS.leftAxis} ${import_timeseries_grid_area_rules.AREAS.leftThreshold} plot ${import_timeseries_grid_area_rules.AREAS.rightThreshold} ${import_timeseries_grid_area_rules.AREAS.rightAxis}"
  ". . ${import_timeseries_grid_area_rules.AREAS.bottomAxis} . ."
`;
const timeseriesGridTemplateColumns = "minmax(0, auto) auto minmax(60%, 1fr) auto minmax(0, auto)";
const TimeseriesChartInternal = /* @__PURE__ */ (0, import_react.forwardRef)(({ config }, forwardedRef) => {
  const loading = (0, import_useChartLayoutLoading.useChartLayoutLoading)();
  const annotations = (0, import_useAnnotationsConfig.useAnnotationsConfig)();
  const tracksQuantity = annotations.tracks.length;
  const numHiddenTracks = annotations.tracks.filter(
    (value) => value.hidden
  ).length;
  const chartRef = (0, import_react.useRef)(null);
  const mergedRef = (0, import_useMergeRefs.useMergeRefs)([
    forwardedRef,
    chartRef
  ]);
  const { width, height } = (0, import_useGraphSize.useGraphSize)();
  const { orientation } = (0, import_useChartLayoutOrientation.useChartLayoutOrientation)();
  const { right: axisRightConfig, left: axisLeftConfig } = (0, import_useTimeseriesFilteredYAxis.useTimeseriesFilteredYAxis)();
  const { units } = (0, import_useTimeseriesFilteredData.useTimeseriesFilteredData)();
  const { hidden: legendHidden } = (0, import_useTimeseriesLegend.useTimeseriesLegend)();
  const annotationsVisibleArea = (0, import_get_annotations_visible_area.getAnnotationsVisibleArea)(
    tracksQuantity,
    annotations.visibleTracksLimit,
    numHiddenTracks
  );
  const timeseriesGridTemplateRows = `minmax(0, ${annotationsVisibleArea + (tracksQuantity && import_constants.ANNOTATIONS_PADDING_BOTTOM)}px)  minmax(60%, 1fr) minmax(0, auto)`;
  const availableWidth = tracksQuantity && orientation === "vertical" ? width - import_constants3.SCROLLBAR_WIDTH : width;
  const isLegendHiddenOrVertical = orientation === "vertical" || legendHidden;
  const isLegendHiddenOrRight = orientation === "horizontal" || legendHidden;
  const bothConfigsUndefinedAndOneUnit = !axisLeftConfig && !axisRightConfig && units.length === 1 && isLegendHiddenOrVertical;
  const bothConfigsSetAndOneUnit = axisLeftConfig && axisRightConfig && units.length === 1 && isLegendHiddenOrVertical;
  const leftConfigSetAndVertical = axisLeftConfig && isLegendHiddenOrVertical;
  const rightConfigSet = axisRightConfig;
  const paddingRight = bothConfigsUndefinedAndOneUnit || bothConfigsSetAndOneUnit || leftConfigSetAndVertical;
  const onlyRightAxis = !axisLeftConfig && rightConfigSet;
  const hasLeftPadding = onlyRightAxis;
  const hasRightPadding = hasLeftPadding || paddingRight;
  const chartGridHeight = height - (!tracksQuantity || isLegendHiddenOrRight ? import_constants4.OVERLAY_PADDING + import_constants4.BOTTOM_PADDING_ADJUSTMENT : 0);
  (0, import_useChartInteractions.useChartInteractions)();
  if (!height || !width) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      style: {
        paddingTop: tracksQuantity ? 0 : import_constants4.OVERLAY_PADDING,
        paddingLeft: hasLeftPadding ? import_constants4.OVERLAY_PADDING : 0
      },
      className: import_defaultChartValues_css.defaultChartSize,
      "aria-label": import_constants2.TIMESERIES_ARIA_LABEL,
      role: "img",
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_ChartGridRoot.ChartGrid,
        {
          width: availableWidth - (hasRightPadding ? import_constants4.OVERLAY_PADDING : 0),
          height: chartGridHeight,
          gridTemplateAreas: timeseriesGridTemplateAreas,
          gridTemplateColumns: timeseriesGridTemplateColumns,
          gridTemplateRows: timeseriesGridTemplateRows,
          ref: mergedRef,
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TimeseriesChartProviders.TimeseriesChartProviders, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
            import_TimeseriesChartImperativeHandler.TimeseriesChartImperativeHandler,
            {
              config,
              forwardedRef,
              chartRef,
              children: [
                /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TimeseriesValueLeftAxisArea.TimeseriesValueLeftAxisArea, {}),
                /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TimeseriesTimeBottomAxisArea.TimeseriesTimeBottomAxisArea, {}),
                /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_ChartGridRoot.ChartGrid.PlotArea, { children: [
                  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TimeseriesChartPlot.TimeseriesChartPlot, {}),
                  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TimeseriesAnnotationsIndicators.TimeseriesAnnotationsIndicators, {}),
                  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                    import_ChartLoaderOverlay.ChartLoaderOverlay,
                    {
                      loading: !!loading,
                      zoomEnabled: !!config.interactions?.zoom?.enabled,
                      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TimeseriesChartOverlays.TimeseriesChartOverlays, {})
                    }
                  ),
                  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ThresholdIndicators.ThresholdIndicators, { indicatorType: "line" })
                ] }),
                /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TimeseriesValueRightAxisArea.TimeseriesValueRightAxisArea, {}),
                /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TimeseriesThresholdArea.TimeseriesThresholdArea, {}),
                !!tracksQuantity && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TimeseriesChartAnnotationsArea.TimeseriesChartAnnotationsArea, {})
              ]
            }
          ) })
        }
      )
    }
  );
});
TimeseriesChartInternal.displayName = import_constants2.DISPLAY_NAME;
