// @dynatrace/strato-components/overlays/hooks/useOverlayWithTrigger.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useOverlayWithTrigger_exports = {};
__export(useOverlayWithTrigger_exports, {
  useOverlayWithTrigger: () => useOverlayWithTrigger
});
module.exports = __toCommonJS(useOverlayWithTrigger_exports);
var import_core = require("@popperjs/core");
var import_react = require("react");
var import_react_popper = require("react-popper");
var import_useDialog = require("./useDialog.js");
var import_useKeyboardInteractionOutside = require("./useKeyboardInteractionOutside.js");
var import_useOverlay = require("./useOverlay.js");
var import_usePress = require("./usePress.js");
var import_useStateRef = require("./useStateRef.js");
var import_useId = require("../../core/hooks/useId.js");
var import_useOverlayStacking = require("../../core/hooks/useOverlayStacking.js");
var import_get_first_focusable_child = require("../../core/utils/focus-management/get-first-focusable-child.js");
var import_get_last_focusable_child = require("../../core/utils/focus-management/get-last-focusable-child.js");
var import_merge_props = require("../../core/utils/merge-props.js");
var import_shared_types = require("../../forms/shared-types.js");
var import_popper_modifiers = require("../utils/popper-modifiers.js");
function getAriaHasPopup(type) {
  if (type === "menu") {
    return true;
  } else if (type === "listbox") {
    return "listbox";
  }
  return void 0;
}
function useOverlayWithTrigger(props = {}) {
  const {
    placement = "bottom",
    placementFallbacks = ["top"],
    offset = 0,
    crossOffset = 0,
    additionalElevation,
    disabled = false,
    widthStrategy = "content",
    heightStrategy = "content",
    trapFocus = false,
    defaultOpen = false,
    open,
    onOpenChange,
    shouldCloseOnBlur = false,
    "aria-labelledby": ariaLabelledBy
  } = props;
  const [targetElement, setTargetElement] = (0, import_useStateRef.useStateRef)();
  const [overlayElement, setOverlayElement, overlayRef] = (0, import_useStateRef.useStateRef)();
  const [, setTriggerElement, triggerRef] = (0, import_useStateRef.useStateRef)();
  const overlayContainerProps = (0, import_useOverlayStacking.useOverlayStacking)(
    void 0,
    additionalElevation
  );
  const overlayState = (0, import_useOverlay.useOverlayTriggerState)({
    defaultOpen,
    isOpen: open,
    onOpenChange
  });
  const { isOpen, close, toggle } = overlayState;
  const overlayId = (0, import_useId.useId)();
  const onKeyDown = (0, import_react.useCallback)(
    (e) => {
      if (e.key === "Tab" && isOpen) {
        const overlay = document.getElementById(`${overlayId}`);
        if (overlay) {
          const firstFocusableChild = (0, import_get_first_focusable_child.getFirstFocusableChild)(overlay);
          const lastFocusableChild = (0, import_get_last_focusable_child.getLastFocusableChild)(overlay);
          if (!firstFocusableChild) {
            e.preventDefault();
            close();
            return;
          }
          const isFirstFocusable = firstFocusableChild === e.target;
          const isLastFocusable = lastFocusableChild === e.target;
          if (!trapFocus && (e.shiftKey && isFirstFocusable || !e.shiftKey && isLastFocusable)) {
            e.preventDefault();
            close();
          }
        }
      }
    },
    [close, isOpen, overlayId, trapFocus]
  );
  const triggerProps = {
    "aria-haspopup": getAriaHasPopup("dialog"),
    "aria-expanded": isOpen,
    "aria-controls": isOpen ? overlayId : void 0,
    onKeyDown
  };
  const { pressProps } = (0, import_usePress.usePress)({
    onPress: toggle,
    isDisabled: disabled
  });
  const { styles: positioningStyles, attributes: positioningAttributes } = (0, import_react_popper.usePopper)(targetElement, overlayElement, {
    placement,
    modifiers: [
      {
        name: "offset",
        options: {
          offset: [crossOffset, offset]
        }
      },
      {
        name: "flip",
        options: {
          fallbackPlacements: [placement, ...placementFallbacks]
        }
      },
      import_popper_modifiers.overlayWidthStrategyHandling,
      {
        name: "overlayWidthStrategyHandling",
        enabled: widthStrategy !== "content",
        options: { widthStrategy }
      },
      import_popper_modifiers.overlayHeightStrategyHandling,
      {
        name: "overlayHeightStrategyHandling",
        enabled: heightStrategy !== "content",
        options: { heightStrategy }
      },
      import_popper_modifiers.observeOverlaySize,
      import_popper_modifiers.observeAnimationDone,
      import_core.eventListeners
    ]
  });
  const { overlayProps } = (0, import_useOverlay.useOverlay)(
    {
      isOpen,
      isDismissible: true,
      shouldCloseOnInteractOutside: (element) => {
        const triggerElement = (0, import_shared_types.isFormControlRef)(triggerRef) ? triggerRef.element : triggerRef;
        if (triggerElement === null) {
          return true;
        }
        return element !== triggerRef.current && triggerRef.current?.contains(element) === false;
      },
      onClose: () => {
        close();
      },
      shouldCloseOnBlur
    },
    overlayRef
  );
  const { dialogProps } = (0, import_useDialog.useDialog)({});
  const { "aria-labelledby": _unused, ...dialogPropsWithoutLabelledBy } = dialogProps;
  (0, import_useKeyboardInteractionOutside.useKeyboardInteractionOutside)({
    ref: overlayRef,
    isDisabled: !isOpen,
    onInteractOutside: (event) => {
      if (event.key === "Escape") {
        close();
      }
    }
  });
  const updateTriggerElement = (0, import_react.useCallback)(
    (triggerRef2) => {
      const newTriggerElement = (0, import_shared_types.isFormControlRef)(triggerRef2) ? triggerRef2.element : triggerRef2;
      setTriggerElement(newTriggerElement);
    },
    [setTriggerElement]
  );
  const updateTargetElement = (0, import_react.useCallback)(
    (targetRef) => {
      const newTargetElement = (0, import_shared_types.isFormControlRef)(targetRef) ? targetRef.element : targetRef;
      setTargetElement(newTargetElement);
    },
    [setTargetElement]
  );
  return {
    triggerRef: updateTriggerElement,
    targetRef: updateTargetElement,
    overlayRef: setOverlayElement,
    overlayProps: {
      ...(0, import_merge_props.mergeProps)(
        overlayProps,
        { ...dialogPropsWithoutLabelledBy, "aria-labelledby": ariaLabelledBy },
        // we also need to add the onKeyDown handler in here for closing overlays with interactive content when tabbing and shouldCloseOnBlur is false (default)
        // if shouldCloseOnBlur is set to true, the focus handlers from the returned useOverlay overlayProps take care of the closing
        // if there's no interactive content, the trigger keeps the focus, so the onKeyDown event from the triggerProps takes care of closing
        { onKeyDown: shouldCloseOnBlur ? void 0 : onKeyDown },
        { style: positioningStyles.popper, id: overlayId, trapFocus },
        positioningAttributes.popper
      )
    },
    overlayContainerProps,
    overlayState,
    overlayTriggerProps: (0, import_merge_props.mergeProps)(pressProps, triggerProps, {
      type: "button",
      disabled
    })
  };
}
