// @dynatrace/strato-components/overlays/hooks/usePress.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var usePress_exports = {};
__export(usePress_exports, {
  usePress: () => usePress
});
module.exports = __toCommonJS(usePress_exports);
var import_react = require("react");
var import_focus_without_scrolling = require("../../buttons/button/utils/focus-without-scrolling.js");
var import_is_virtual_event = require("../../buttons/button/utils/is-virtual-event.js");
const modifiedElementMap = /* @__PURE__ */ new WeakMap();
function disableTextSelection(target) {
  if (target instanceof HTMLElement || target instanceof SVGElement) {
    modifiedElementMap.set(target, target.style.userSelect);
    target.style.userSelect = "none";
  }
}
function restoreTextSelection(target) {
  if (target instanceof HTMLElement || target instanceof SVGElement) {
    if (target && modifiedElementMap.has(target)) {
      const targetOldUserSelect = modifiedElementMap.get(target);
      if (target.style.userSelect === "none") {
        target.style.userSelect = targetOldUserSelect;
      }
      if (target.getAttribute("style") === "") {
        target.removeAttribute("style");
      }
      modifiedElementMap.delete(target);
    }
  }
}
const nonTextInputTypes = /* @__PURE__ */ new Set([
  "checkbox",
  "radio",
  "range",
  "color",
  "file",
  "image",
  "button",
  "submit",
  "reset"
]);
function useGlobalListeners() {
  const globalListeners = (0, import_react.useRef)(/* @__PURE__ */ new Map());
  const addGlobalListener = (0, import_react.useCallback)(
    (eventTarget, type, listener, options) => {
      const fn = options?.once ? (
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        (...args) => {
          globalListeners.current.delete(listener);
          listener({ ...args });
        }
      ) : listener;
      globalListeners.current.set(listener, { type, eventTarget, fn, options });
      eventTarget.addEventListener(type, listener, options);
    },
    []
  );
  const removeGlobalListener = (0, import_react.useCallback)(
    (eventTarget, type, listener, options) => {
      const fn = globalListeners.current.get(listener)?.fn || listener;
      eventTarget.removeEventListener(type, fn, options);
      globalListeners.current.delete(listener);
    },
    []
  );
  const removeAllGlobalListeners = (0, import_react.useCallback)(() => {
    globalListeners.current.forEach((value, key) => {
      removeGlobalListener(value.eventTarget, value.type, key, value.options);
    });
  }, [removeGlobalListener]);
  (0, import_react.useEffect)(() => {
    return removeAllGlobalListeners;
  }, [removeAllGlobalListeners]);
  return { addGlobalListener, removeAllGlobalListeners };
}
function isValidInputKey(target, key) {
  return target.type === "checkbox" || target.type === "radio" ? key === " " : nonTextInputTypes.has(target.type);
}
function isHTMLAnchorLink(target) {
  return target.tagName === "A" && target.hasAttribute("href");
}
function isValidKeyboardEvent(event, currentTarget) {
  const { key, code } = event;
  const element = currentTarget;
  const role = element.getAttribute("role");
  return (key === "Enter" || key === " " || code === "Space") && !(element instanceof HTMLInputElement && !isValidInputKey(element, key) || element instanceof HTMLTextAreaElement || element.isContentEditable) && // A link with a valid href should be handled natively,
  // unless it also has role='button' and was triggered using Space.
  (!isHTMLAnchorLink(element) || role === "button" && key !== "Enter") && // An element with role='link' should only trigger with Enter key
  !(role === "link" && key !== "Enter");
}
function shouldPreventDefault(target) {
  return !(target instanceof HTMLElement) || !target.draggable;
}
function shouldPreventDefaultKeyboard(target, key) {
  if (target instanceof HTMLInputElement) {
    return !isValidInputKey(target, key);
  }
  if (target instanceof HTMLButtonElement) {
    return target.type !== "submit";
  }
  return true;
}
function getPointClientRect(point) {
  const pointOffsetX = point.width !== void 0 ? point.width / 2 : point.radiusX;
  const pointOffsetY = point.height !== void 0 ? point.height / 2 : point.radiusY;
  const offsetX = pointOffsetX || 0;
  const offsetY = pointOffsetY || 0;
  return {
    top: point.clientY - offsetY,
    right: point.clientX + offsetX,
    bottom: point.clientY + offsetY,
    left: point.clientX - offsetX
  };
}
function areRectanglesOverlapping(a, b) {
  if (a.left > b.right || b.left > a.right) {
    return false;
  }
  if (a.top > b.bottom || b.top > a.bottom) {
    return false;
  }
  return true;
}
function isOverTarget(point, target) {
  const rect = target.getBoundingClientRect();
  const pointRect = getPointClientRect(point);
  return areRectanglesOverlapping(rect, pointRect);
}
function usePress(props) {
  const { onPress, isDisabled } = props;
  const propsRef = (0, import_react.useRef)({
    onPress,
    isDisabled
  });
  (0, import_react.useEffect)(() => {
    propsRef.current = {
      onPress,
      isDisabled
    };
  }, [isDisabled, onPress]);
  const pressStateRef = (0, import_react.useRef)({
    isPressed: false,
    ignoreEmulatedMouseEvents: false,
    ignoreClickAfterPress: false,
    didFirePressStart: false,
    activePointerId: null,
    target: null,
    isOverTarget: false,
    pointerType: null
  });
  const { addGlobalListener, removeAllGlobalListeners } = useGlobalListeners();
  const pressProps = (0, import_react.useMemo)(() => {
    const state = pressStateRef.current;
    let stateTarget = state.target;
    const triggerPressStart = () => {
      const { isDisabled: isDisabled2 } = propsRef.current;
      if (isDisabled2 || state.didFirePressStart) {
        return;
      }
      state.didFirePressStart = true;
    };
    const triggerPressEnd = (wasPressed = true) => {
      const { onPress: onPress2, isDisabled: isDisabled2 } = propsRef.current;
      if (!state.didFirePressStart) {
        return;
      }
      state.ignoreClickAfterPress = true;
      state.didFirePressStart = false;
      if (onPress2 && wasPressed && !isDisabled2) {
        onPress2();
      }
    };
    const triggerPressUp = () => {
      const { isDisabled: isDisabled2 } = propsRef.current;
      if (isDisabled2) {
        return;
      }
    };
    const cancel = () => {
      if (state.isPressed) {
        if (state.isOverTarget) {
          triggerPressEnd(false);
        }
        state.isPressed = false;
        state.isOverTarget = false;
        state.activePointerId = null;
        state.pointerType = null;
        removeAllGlobalListeners();
        restoreTextSelection(stateTarget);
      }
    };
    const pressProps2 = {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      onKeyDown(e) {
        if (e.metaKey || e.ctrlKey || e.altKey) {
          return;
        }
        const currentTarget = e.currentTarget;
        if (isValidKeyboardEvent(e.nativeEvent, currentTarget) && isValidKeyboardEvent(e, currentTarget) && currentTarget.contains(e.target)) {
          if (shouldPreventDefaultKeyboard(e.target, e.key)) {
            e.preventDefault();
          }
          e.stopPropagation();
          if (!state.isPressed && !e.repeat) {
            state.target = e.currentTarget;
            stateTarget = state.target;
            state.isPressed = true;
            state.pointerType = "keyboard";
            triggerPressStart();
            addGlobalListener(document, "keyup", onKeyUp, false);
          }
        } else if (e.key === "Enter" && isHTMLAnchorLink(e.currentTarget)) {
          e.stopPropagation();
        }
      },
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      onKeyUp(e) {
        const currentTarget = e.currentTarget;
        if (isValidKeyboardEvent(e.nativeEvent, currentTarget) && !e.repeat && currentTarget.contains(e.target)) {
          triggerPressUp();
        }
      },
      onClick(e) {
        const currentTarget = e.currentTarget;
        if (e && !currentTarget.contains(e.target)) {
          return;
        }
        if (e && e.button === 0) {
          e.stopPropagation();
          if (isDisabled) {
            e.preventDefault();
          }
          if (!state.ignoreClickAfterPress && !state.ignoreEmulatedMouseEvents && (state.pointerType === "virtual" || (0, import_is_virtual_event.isVirtualClick)(e.nativeEvent))) {
            if (!isDisabled) {
              (0, import_focus_without_scrolling.focusWithoutScrolling)(currentTarget);
            }
            triggerPressStart();
            triggerPressUp();
            triggerPressEnd();
          }
          state.ignoreEmulatedMouseEvents = false;
          state.ignoreClickAfterPress = false;
        }
      }
    };
    const onKeyUp = (e) => {
      if (state.isPressed && isValidKeyboardEvent(e, stateTarget)) {
        if (shouldPreventDefaultKeyboard(e.target, e.key)) {
          e.preventDefault();
        }
        e.stopPropagation();
        state.isPressed = false;
        const target = e.target;
        triggerPressEnd(stateTarget.contains(target));
        removeAllGlobalListeners();
        if (state.target instanceof HTMLElement && state.target.contains(target) && (isHTMLAnchorLink(state.target) || state.target.getAttribute("role") === "link")) {
          state.target.click();
        }
      }
    };
    pressProps2.onPointerDown = (e) => {
      if (e.button !== 0 || !e.currentTarget.contains(e.target)) {
        return;
      }
      if ((0, import_is_virtual_event.isVirtualPointerEvent)(e.nativeEvent)) {
        state.pointerType = "virtual";
        return;
      }
      if (shouldPreventDefault(e.currentTarget)) {
        e.preventDefault();
      }
      state.pointerType = e.pointerType;
      if (!state.isPressed) {
        const currentTarget = e.currentTarget;
        state.isPressed = true;
        state.isOverTarget = true;
        state.activePointerId = e.pointerId;
        state.target = currentTarget;
        stateTarget = state.target;
        if (!isDisabled) {
          (0, import_focus_without_scrolling.focusWithoutScrolling)(currentTarget);
        }
        disableTextSelection(stateTarget);
        triggerPressStart();
        addGlobalListener(document, "pointerup", onPointerUp, false);
        addGlobalListener(
          document,
          "pointercancel",
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          onPointerCancel,
          false
        );
      }
    };
    pressProps2.onMouseDown = (e) => {
      if (!e.currentTarget.contains(e.target)) {
        return;
      }
      if (e.button === 0) {
        if (shouldPreventDefault(e.currentTarget)) {
          e.preventDefault();
        }
        e.stopPropagation();
      }
    };
    pressProps2.onPointerUp = (e) => {
      if (!e.currentTarget.contains(e.target) || state.pointerType === "virtual") {
        return;
      }
      if (e.button === 0 && isOverTarget(e, e.currentTarget)) {
        triggerPressUp();
      }
    };
    const onPointerUp = (e) => {
      if (e.pointerId === state.activePointerId && state.isPressed && e.button === 0) {
        if (isOverTarget(e, stateTarget)) {
          triggerPressEnd();
        } else if (state.isOverTarget) {
          triggerPressEnd(false);
        }
        state.isPressed = false;
        state.isOverTarget = false;
        state.activePointerId = null;
        state.pointerType = null;
        removeAllGlobalListeners();
        restoreTextSelection(stateTarget);
      }
    };
    const onPointerCancel = () => {
      cancel();
    };
    pressProps2.onDragStart = (e) => {
      if (!e.currentTarget.contains(e.target)) {
        return;
      }
      cancel();
    };
    return pressProps2;
  }, [addGlobalListener, isDisabled, removeAllGlobalListeners]);
  return {
    pressProps
  };
}
