// @dynatrace/strato-components/forms/select/Select.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Select_exports = {};
__export(Select_exports, {
  Select: () => Select
});
module.exports = __toCommonJS(Select_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_select = __toESM(require("react-select"));
var import_components = require("./components/components.js");
var import_SelectMenuContext = require("./contexts/SelectMenuContext.js");
var import_use_select_options = require("./hooks/use-select-options.js");
var import_use_select_slots = require("./hooks/use-select-slots.js");
var import_use_select_value = require("./hooks/use-select-value.js");
var import_SelectContent = require("./SelectContent.js");
var import_SelectCustomTrigger = require("./SelectCustomTrigger.js");
var import_SelectDisplayValue = require("./SelectDisplayValue.js");
var import_SelectEmptyState = require("./SelectEmptyState.js");
var import_SelectFilter = require("./SelectFilter.js");
var import_SelectGroup = require("./SelectGroup.js");
var import_SelectOption = require("./SelectOption.js");
var import_SelectPrefix = require("./SelectPrefix.js");
var import_SelectSuffix = require("./SelectSuffix.js");
var import_SelectTrigger = require("./SelectTrigger.js");
var import_select_menu_state_reducer = require("./state/select-menu-state-reducer.js");
var styles = __toESM(require("./styles/Select.sty.js"));
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useControllableState = require("../../core/hooks/useControllableState.js");
var import_useFormControlState = require("../../core/hooks/useFormControlState.js");
var import_sdk_logger = require("../../core/sdk/sdk-logger.js");
var import_merge_props = require("../../core/utils/merge-props.js");
var import_FormFieldMessageContextProvider = require("../forms-core/components/form-field/FormFieldMessageContextProvider.js");
var import_useFormFieldMessagesContext = require("../forms-core/hooks/useFormFieldMessagesContext.js");
var import_useFormFieldValidation = require("../forms-core/hooks/useFormFieldValidation.js");
var import_useSyncedFormControlProps = require("../forms-core/hooks/useSyncedFormControlProps.js");
var import_formfield_messages_state_reducer = require("../forms-core/state/formfield-messages-state-reducer.js");
const COMPONENT_NAME = "Select";
function getNewFocusedOption(options, focusedOptionValue) {
  for (const optionOrGroup of options) {
    if ((0, import_use_select_options._isOption)(optionOrGroup)) {
      if (optionOrGroup.value === focusedOptionValue) {
        return optionOrGroup;
      }
    } else {
      for (const option of optionOrGroup.options) {
        if ((0, import_use_select_options._isOption)(option) && option.value === focusedOptionValue) {
          return option;
        }
      }
    }
  }
  return null;
}
function SelectFwd(props, forwardedRef) {
  const {
    multiple = false,
    disabled: propsDisabled,
    controlState,
    clearable = multiple,
    name,
    id: propId,
    readOnly: propReadOnly,
    children,
    value: propsValue,
    defaultValue: propsDefaultValue,
    onChange: propsOnChange,
    onFocus,
    onBlur,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "aria-label": ariaLabel,
    "aria-labelledby": ariaLabelledBy,
    "aria-describedby": ariaDescribedBy,
    "aria-details": ariaDetails,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "aria-disabled": ariaDisabledProp,
    onOpenChange,
    required: propsRequired,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const slots = (0, import_use_select_slots.useSelectSlots)(children);
  const stratoLogger = (0, import_react.useMemo)(
    () => import_sdk_logger.createStratoLogger.bind({
      name,
      id: propId,
      ariaLabel,
      dataTestId
    })(COMPONENT_NAME),
    [name, propId, ariaLabel, dataTestId]
  );
  const {
    disableFiltering,
    value: filterValueProp,
    defaultValue: filterDefaultValueProp,
    onChange: filterOnChangeProp,
    ...remainingFilterProps
  } = slots.Filter?.props ?? {};
  const {
    width: triggerWidthProp,
    placeholder: triggerPlaceholderProp,
    ...remainingTriggerProps
  } = slots.Trigger?.props ?? {};
  const {
    width: contentWidthProp,
    loading: contentLoadingProp,
    showSelectedOptionsFirst: contentShowSelectedOptionsFirstProp,
    ...remainingContentProps
  } = slots.Content?.props ?? {};
  const {
    optionsOrGroups: options,
    optionsHavePrefix,
    allOptions
  } = (0, import_use_select_options.useSelectOptions)(slots.Content, stratoLogger);
  const { value, defaultValue } = (0, import_use_select_value.useSelectValue)({
    value: propsValue,
    defaultValue: propsDefaultValue,
    options,
    multiple: multiple === true
  });
  const hiddenSelectRef = (0, import_react.useRef)(null);
  const [triggerRef, setTriggerRef] = (0, import_react.useState)(null);
  const { disabled, required, id, ariaDisabled, readOnly } = (0, import_useSyncedFormControlProps.useSyncedFormControlProps)({
    propDisabled: propsDisabled,
    propRequired: propsRequired,
    propId,
    propAriaDisabled: ariaDisabledProp,
    propReadOnly
  });
  const { onBlur: validationOnBlur } = (0, import_useFormFieldValidation.useFormFieldValidation)(hiddenSelectRef, {
    required
  });
  const menuOpenReducer = (0, import_react.useCallback)(
    (state, action) => {
      if (readOnly) {
        return state;
      }
      const newState = (0, import_select_menu_state_reducer.reducer)(state, action);
      if (state !== newState) {
        queueMicrotask(() => {
          newState === false && !controlState && validationOnBlur();
          onOpenChange?.(newState);
        });
      }
      return newState;
    },
    [readOnly, controlState, validationOnBlur, onOpenChange]
  );
  const [menuOpen, dispatch] = (0, import_react.useReducer)(menuOpenReducer, false);
  const [isFocusWithin, setIsFocusWithin] = (0, import_react.useState)(false);
  const [filterValue, setFilterValue] = (0, import_useControllableState.useControllableState)({
    value: filterValueProp,
    defaultValue: filterDefaultValueProp,
    onChange: filterOnChangeProp
  });
  const { state: messagesContextState, dispatch: messagesContextDispatch } = (0, import_useFormFieldMessagesContext.useFormFieldMessagesContext)();
  const selectRef = (0, import_react.useRef)(null);
  const hiddenTrigger = (0, import_react.useMemo)(
    () => triggerRef?.querySelector("input"),
    [triggerRef]
  );
  const [focusOnOption, setFocusOnOption] = (0, import_react.useState)(true);
  (0, import_react.useEffect)(() => {
    if (disableFiltering === true && (filterValue !== void 0 || filterValue !== "")) {
      setFilterValue("");
    }
  }, [disableFiltering]);
  const filterConfig = {
    // Filter by both visible label and option value/textValue.
    // This allows searching by value (e.g. "green") when the rendered label differs.
    stringify: (option) => {
      const searchValue = option.data.textValue ?? option.value;
      if (searchValue && searchValue !== option.label) {
        return `${option.label} ${searchValue}`;
      }
      return `${option.label}`;
    }
  };
  const handleChange = (option) => {
    requestAnimationFrame(() => {
      const nativeEvent = new Event("change", {
        bubbles: true,
        cancelable: true
      });
      hiddenSelectRef?.current?.dispatchEvent(nativeEvent);
    });
    if (multiple === false) {
      setFilterValue("");
    }
    if (!propsOnChange) {
      return;
    }
    if (multiple === false) {
      if (option === value) {
        return;
      }
      if (option === null) {
        propsOnChange?.(null);
      } else {
        propsOnChange?.(
          option?.value
        );
      }
    }
    if (multiple === true) {
      const selectedOptionValues = option.map((individualOption) => individualOption.value);
      propsOnChange?.(
        selectedOptionValues
      );
    }
  };
  const filterRef = (0, import_react.useRef)(null);
  const clearSelectionRef = (0, import_react.useRef)(null);
  const containerRef = (0, import_react.useRef)(null);
  const contentRef = (0, import_react.useRef)(null);
  const menuListFunctionRef = (0, import_react.useRef)(null);
  const handleArrowKeyOptionFocus = (key) => {
    if (key !== "ArrowUp" && key !== "ArrowDown") {
      return;
    }
    requestAnimationFrame(() => {
      if (!contentRef.current) {
        return;
      }
      const container = contentRef.current;
      const focusedElement = container.querySelector(
        ".strato-select-option--focus"
      );
      if (focusedElement) {
        focusedElement.scrollIntoView({
          block: "nearest"
        });
      } else if (container) {
        if (key === "ArrowDown") {
          container.scrollTop = 0;
        } else if (menuListFunctionRef.current) {
          menuListFunctionRef.current.scrollToBottom();
        }
      }
    });
  };
  (0, import_react.useImperativeHandle)(
    forwardedRef,
    () => ({
      element: containerRef.current,
      triggerRef,
      inputRef: hiddenSelectRef.current,
      focus: () => {
        hiddenTrigger?.focus();
      },
      open: () => {
        hiddenTrigger?.focus();
        dispatch({
          type: import_select_menu_state_reducer.SelectActions.OPEN
        });
      },
      close: () => dispatch({
        type: import_select_menu_state_reducer.SelectActions.CLOSE
      }),
      validate: () => {
        queueMicrotask(() => hiddenSelectRef.current?.reportValidity());
        messagesContextDispatch?.({
          type: import_formfield_messages_state_reducer.FormFieldMessagesAction.SETISTOUCHED,
          payload: { isTouched: true }
        });
      }
    }),
    [hiddenTrigger, messagesContextDispatch, triggerRef]
  );
  if (multiple && propsOnChange) {
    const select = selectRef.current;
    if (select?.state.focusedOption) {
      const focusedOptionValue = select.state.focusedOption.value;
      const newFocusedOption = getNewFocusedOption(options, focusedOptionValue);
      select.state.focusedOption = newFocusedOption;
    }
  }
  const [hasFormControlError] = (0, import_useFormControlState.useFormControlState)(controlState);
  const showError = controlState ? hasFormControlError : messagesContextState.showError;
  const contextMemo = (0, import_react.useMemo)(
    () => ({ state: menuOpen, dispatch }),
    [menuOpen, dispatch]
  );
  const onBlurHandler = (event) => {
    if (
      // The relatedTarget is null when the focus leaves the browser and we want to still trigger the onBlur in this case.
      !event.relatedTarget || event.relatedTarget !== hiddenTrigger && event.relatedTarget !== clearSelectionRef.current && event.relatedTarget !== filterRef.current?.inputRef
    ) {
      if (!menuOpen && isFocusWithin) {
        setIsFocusWithin(false);
        onBlur?.(event);
      }
    }
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_SelectMenuContext.SelectMenuContext.Provider, { value: contextMemo, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      ref: containerRef,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      className: (0, import_clsx.default)(
        styles.selectWrapper({
          fullWidth: triggerWidthProp === "full" || triggerWidthProp === "100%"
        }),
        consumerClassName
      ),
      style: consumerStyle,
      ...(0, import_merge_props.mergeProps)(restProps, behavioralTrackingProps),
      "data-dt-component": COMPONENT_NAME,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_react_select.default,
        {
          instanceId: id,
          selectInputId: id,
          className: (0, import_clsx.default)(styles.select, styles.clearReactSelectZIndex, {
            [styles.validAndEnabled]: !disabled && !ariaDisabled && !showError
          }),
          unstyled: true,
          options,
          components: import_components.components,
          value,
          defaultValue,
          onChange: (value2) => (
            // Need to cast this here as React.Select handles the value still as unknown.
            // TODO: Maybe we can go to the JSX generic?
            handleChange(
              value2
            )
          ),
          onBlur: onBlurHandler,
          hiddenSelectName: name,
          ref: selectRef,
          pageSize: 5,
          "aria-label": ariaLabel,
          "aria-labelledby": ariaLabelledBy,
          "aria-describedby": ariaDescribedBy,
          "aria-details": ariaDetails,
          "aria-disabled": ariaDisabled,
          hiddenSelectRef,
          isSearchable: false,
          isMulti: multiple ?? false,
          hideSelectedOptions: false,
          closeMenuOnSelect: multiple === false,
          isClearable: false,
          isDisabled: disabled,
          isRequired: required,
          triggerWidth: triggerWidthProp,
          contentWidth: contentWidthProp ?? "trigger",
          placeholderText: triggerPlaceholderProp,
          displayValue: slots.DisplayValue ?? void 0,
          prefix: slots.Prefix ?? void 0,
          customTrigger: slots.CustomTrigger ?? void 0,
          triggerRef,
          setTriggerRef,
          optionsHavePrefix,
          menuIsOpen: menuOpen,
          filterOption: (option, inputValue) => {
            if (disableFiltering === true) {
              return true;
            }
            const filterFunction = (0, import_react_select.createFilter)(filterConfig);
            return filterFunction(
              option,
              inputValue
            );
          },
          customFilter: slots.Filter ?? void 0,
          inputValue: filterValue,
          onInputChange: (newValue, actionmeta) => {
            if (actionmeta.action === "input-change") {
              setFilterValue(newValue);
            }
          },
          emptyState: slots.EmptyState ?? void 0,
          controlState,
          loading: contentLoadingProp,
          clearable,
          allOptions,
          tabSelectsValue: false,
          focusOnOption,
          contentRef,
          menuListFunctionRef,
          filterRef,
          clearSelectionRef,
          showSelectedOptionsFirst: contentShowSelectedOptionsFirstProp,
          componentProps: {
            trigger: {
              className: slots.Trigger?.props.className,
              style: { ...slots.Trigger?.props.style },
              readOnly: !!readOnly,
              ariaDisabled,
              ...remainingTriggerProps
            },
            filter: {
              className: slots.Filter?.props.className,
              style: slots.Filter?.props.style,
              ...remainingFilterProps
            },
            content: {
              className: slots.Content?.props.className,
              style: slots.Content?.props.style,
              ...remainingContentProps
            },
            emptyState: {
              ...slots.EmptyState?.props
            }
          },
          onFocus: (event) => {
            if (!isFocusWithin) {
              onFocus?.(event);
              setIsFocusWithin(true);
            }
          },
          onKeyDown: (event) => {
            if (!menuOpen) {
              if (event.key === "Enter" || event.key === " ") {
                dispatch({
                  type: import_select_menu_state_reducer.SelectActions.SELECTFOCUS
                });
              }
              return true;
            }
            if (event.key === "Escape") {
              dispatch({
                type: import_select_menu_state_reducer.SelectActions.SELECTFOCUS
              });
              hiddenTrigger?.focus();
            }
            if (filterRef.current || clearSelectionRef.current) {
              if (document.activeElement === filterRef.current?.inputRef) {
                if (event.key === "Backspace" && filterRef.current.inputRef?.value === "") {
                  event.preventDefault();
                  event.stopPropagation();
                }
                if (event.key === "Tab") {
                  if (!clearSelectionRef.current || event.shiftKey) {
                    event.preventDefault();
                    event.stopPropagation();
                    hiddenTrigger?.focus();
                  } else {
                    clearSelectionRef.current && setFocusOnOption(false);
                  }
                }
                if (event.key === "Enter" && !props.multiple) {
                  dispatch({
                    type: import_select_menu_state_reducer.SelectActions.SELECTFOCUS
                  });
                  hiddenTrigger?.focus();
                }
                if (event.key === "Enter" && contentLoadingProp) {
                  event.preventDefault();
                  event.stopPropagation();
                }
                if (event.key === " " && !filterValue) {
                  setFilterValue(" ");
                  event.preventDefault();
                  event.stopPropagation();
                }
                handleArrowKeyOptionFocus(event.key);
                return true;
              }
              if (document.activeElement === clearSelectionRef.current) {
                setFocusOnOption(false);
                if (event.key === "ArrowDown" || event.key === "ArrowUp") {
                  event.preventDefault();
                }
                if (event.key === "Tab" && !(event.shiftKey && filterRef.current)) {
                  event.preventDefault();
                  event.stopPropagation();
                  hiddenTrigger?.focus();
                  setFocusOnOption(true);
                }
                if (event.key === "Tab" && event.shiftKey && filterRef.current) {
                  setFocusOnOption(true);
                }
                return false;
              }
              try {
                if (document.activeElement && document.getElementById(document.activeElement.id)) {
                  if (event.key === "Tab") {
                    event.preventDefault();
                    event.stopPropagation();
                    if (!event.shiftKey) {
                      if (filterRef.current) {
                        filterRef.current.inputRef?.focus();
                        return false;
                      }
                      if (clearSelectionRef.current) {
                        clearSelectionRef.current.focus();
                        setFocusOnOption(false);
                        return false;
                      }
                    } else {
                      if (clearSelectionRef.current) {
                        clearSelectionRef.current.focus();
                        setFocusOnOption(false);
                        return false;
                      }
                      if (filterRef.current) {
                        filterRef.current.inputRef?.focus();
                        return false;
                      }
                    }
                  }
                  handleArrowKeyOptionFocus(event.key);
                  if (contentLoadingProp) {
                    event.preventDefault();
                    event.stopPropagation();
                  }
                }
              } catch (error) {
                console.log(
                  "Invalid query selector on the Select trigger",
                  error
                );
              }
              return false;
            }
            try {
              if (document.activeElement && document.getElementById(document.activeElement.id)) {
                if (event.key === "Tab") {
                  dispatch({
                    type: import_select_menu_state_reducer.SelectActions.SELECTFOCUS
                  });
                } else {
                  handleArrowKeyOptionFocus(event.key);
                }
              }
            } catch (error) {
              console.log(
                "Invalid query selector on the Select trigger",
                error
              );
            }
            setFocusOnOption(true);
          }
        }
      )
    }
  ) });
}
const SelectForwardRef = (0, import_react.forwardRef)(SelectFwd);
const SelectRoot = (0, import_FormFieldMessageContextProvider.withFormFieldMessageContextProviderWithGenerics)(SelectForwardRef);
const Select = Object.assign(SelectRoot, {
  Content: import_SelectContent.SelectContent,
  Option: import_SelectOption.SelectOption,
  Group: import_SelectGroup.SelectGroup,
  GroupLabel: import_SelectGroup.SelectGroupLabel,
  Trigger: import_SelectTrigger.SelectTrigger,
  DisplayValue: import_SelectDisplayValue.SelectDisplayValue,
  Prefix: import_SelectPrefix.SelectPrefix,
  Suffix: import_SelectSuffix.SelectSuffix,
  CustomTrigger: import_SelectCustomTrigger.SelectCustomTrigger,
  Filter: import_SelectFilter.SelectFilter,
  EmptyState: import_SelectEmptyState.SelectEmptyState
});
Select.displayName = COMPONENT_NAME;
