// @dynatrace/strato-components/esm/charts/categorical-bar/components/interactions/CategoricalBarChartOverlays.js
import { jsx, jsxs } from "react/jsx-runtime";
import { useRef, useState } from "react";
import { useCategoricalMouseEventHandlers } from "./hooks/useCategoricalMouseEventHandlers.js";
import { CategoricalBarChartInspectOverlay } from "./overlays/CategoricalBarChartInspectOverlay.js";
import { CategoricalBarChartOverlayRefContext } from "../../contexts/CategoricalBarChartOverlayRefContext.js";
import { CategoricalBarChartToolbar } from "../toolbar/CategoricalBarChartToolbar.js";
const CategoricalBarChartOverlays = () => {
  const [shouldShowInspect, setShouldShowInspect] = useState(true);
  const overlayRef = useRef(null);
  const {
    handleMouseEnter,
    handleMouseMove,
    handleMouseLeave,
    handleFocus,
    handleBlur
  } = useCategoricalMouseEventHandlers();
  const handleToolbarMouseEnter = () => {
    setShouldShowInspect(false);
  };
  const handleToolbarMouseLeave = () => {
    setShouldShowInspect(true);
  };
  return /* @__PURE__ */ jsx(
    "div",
    {
      tabIndex: 0,
      ref: overlayRef,
      onFocus: handleFocus,
      onBlur: handleBlur,
      onMouseEnter: handleMouseEnter,
      onMouseMove: handleMouseMove,
      onMouseLeave: handleMouseLeave,
      style: { height: "100%" },
      children: /* @__PURE__ */ jsxs(CategoricalBarChartOverlayRefContext.Provider, { value: overlayRef, children: [
        shouldShowInspect && /* @__PURE__ */ jsx(CategoricalBarChartInspectOverlay, {}),
        /* @__PURE__ */ jsx(
          CategoricalBarChartToolbar,
          {
            onMouseEnter: handleToolbarMouseEnter,
            onMouseLeave: handleToolbarMouseLeave
          }
        )
      ] })
    }
  );
};
CategoricalBarChartOverlays["displayName"] = "CategoricalBarChartInteractions";
export {
  CategoricalBarChartOverlays
};
