// @dynatrace/strato-components/charts/core/components/sparkline-display/components/ContextValueIndicator.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ContextValueIndicator_exports = {};
__export(ContextValueIndicator_exports, {
  ContextValueIndicator: () => ContextValueIndicator
});
module.exports = __toCommonJS(ContextValueIndicator_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_TextEllipsis = require("../../../../../typography/text-ellipsis/TextEllipsis.js");
var import_constants = require("../../annotations/constants.js");
var import_GridLine = require("../../grid/GridLine.js");
var import_chars_font_width_map = require("../../text-ellipsis/chars-font-width-map.js");
var import_truncate_string = require("../../text-ellipsis/truncate-string.js");
var import_constants2 = require("../constants.js");
var import_Sparkline_css = require("../Sparkline.sty.js");
var import_sparkline_container_height = require("../types/sparkline-container-height.js");
var import_get_context_value_x_offset = require("../utils/get-context-value-x-offset.js");
function getTextOffset(isMax, fontSize) {
  if (isMax) {
    return import_constants2.CONTEXT_VALUE_TEXT_OFFSET_MAX;
  }
  return fontSize === import_sparkline_container_height.SparklineContainerHeightFontSize.SMALLER_THAN_32 ? import_constants2.CONTEXT_VALUE_TEXT_OFFSET_MIN_SMALL : import_constants2.CONTEXT_VALUE_TEXT_OFFSET_MIN_LARGE;
}
function computeIndicatorPositioning(props) {
  const {
    formattedValue,
    yPosition,
    dottedLineWidth,
    containerHeight,
    fontSize,
    labelsAlignment,
    chartLeftOffset,
    labelsFixedWidth,
    labelYPosition,
    isMax = false
  } = props;
  const isNegative = formattedValue.startsWith(import_constants2.NEGATIVE_SIGN);
  const displayValue = isNegative ? formattedValue.slice(1) : formattedValue;
  const isLeft = labelsAlignment !== "right";
  const actualWidth = (0, import_truncate_string.calculateWidth)(
    displayValue,
    import_chars_font_width_map.charsFontWidth12pixelsSize500Map
  );
  const labelWidth = typeof labelsFixedWidth === "number" && !Number.isNaN(labelsFixedWidth) ? labelsFixedWidth : actualWidth;
  const positioningWidth = isLeft ? Math.min(actualWidth, labelWidth) : labelWidth;
  const { signXOffset, textXOffset } = (0, import_get_context_value_x_offset.getContextValueXOffset)(
    isLeft,
    chartLeftOffset,
    positioningWidth,
    dottedLineWidth,
    containerHeight,
    fontSize === import_sparkline_container_height.SparklineContainerHeightFontSize.SMALLER_THAN_32
  );
  const textOffset = getTextOffset(isMax, fontSize);
  const labelYDelta = (labelYPosition ?? yPosition) - yPosition;
  return {
    containerTransform: `translateY(${yPosition}px) translateX(${isLeft ? 0 : chartLeftOffset}px)`,
    svgLeft: isLeft ? chartLeftOffset : 0,
    labelYTransform: `translateY(${labelYDelta - textOffset}px)`,
    signXOffset,
    textXOffset,
    actualWidth,
    isNegative,
    displayValue
  };
}
const ContextValueIndicator = (props) => {
  const { dottedLineWidth, containerHeight, fontSize, color } = props;
  const {
    containerTransform,
    svgLeft,
    labelYTransform,
    signXOffset,
    textXOffset,
    actualWidth,
    isNegative,
    displayValue
  } = computeIndicatorPositioning(props);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    "div",
    {
      className: (0, import_Sparkline_css.contextValueText)({ fontSize }),
      style: { transform: containerTransform },
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          "svg",
          {
            width: dottedLineWidth,
            style: { position: "absolute", left: svgLeft, top: 0 },
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              "line",
              {
                "data-testid": import_constants.TestIds.AnnotationIndicator,
                x1: 0,
                x2: dottedLineWidth,
                y1: import_constants2.CONTEXT_VALUES_Y_OFFSET,
                y2: import_constants2.CONTEXT_VALUES_Y_OFFSET,
                stroke: import_constants2.CONTEXT_VALUE_STROKE_COLOR,
                strokeWidth: import_GridLine.DEFAULT_STROKE_WIDTH,
                strokeDasharray: import_constants2.CONTEXT_VALUE_STROKE_DASHARRAY,
                strokeLinecap: import_constants2.CONTEXT_VALUE_STROKE_LINECAP
              }
            )
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          "div",
          {
            className: (0, import_clsx.default)((0, import_Sparkline_css.negativeSign)({ containerHeight })),
            style: {
              transform: `${labelYTransform} translateX(${signXOffset}px)`
            },
            children: isNegative && import_constants2.NEGATIVE_SIGN
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_TextEllipsis.TextEllipsis,
          {
            truncationMode: "middle",
            style: {
              width: `${actualWidth}px`,
              position: "absolute",
              transform: `${labelYTransform} translateX(${textXOffset}px)`,
              color
            },
            children: displayValue
          }
        )
      ]
    }
  );
};
