// @dynatrace/strato-components/charts/core/components/threshold-bar/markers/CircleMarker.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CircleMarker_exports = {};
__export(CircleMarker_exports, {
  CircleMarker: () => CircleMarker
});
module.exports = __toCommonJS(CircleMarker_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_group = require("@visx/group");
var import_react = require("react");
var import_constants = require("./constants.js");
var import_useId = require("../../../../../core/hooks/useId.js");
var import_useMouseHover = require("../../../hooks/useMouseHover.js");
var import_ThresholdBrightnessFilter = require("../ThresholdBrightnessFilter.js");
const CircleMarker = (props) => {
  const { x, y, color, onMouseMove, onMouseLeave } = props;
  const circleRef = (0, import_react.useRef)(null);
  const hovered = (0, import_useMouseHover.useMouseHover)(circleRef);
  const id = (0, import_useId.useId)();
  const brightnessFilterId = `threshold-circle-marker-filter-${id}`;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_group.Group, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ThresholdBrightnessFilter.ThresholdBrightnessFilter, { id: brightnessFilterId }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "circle",
      {
        role: "graphics-symbol",
        "aria-label": "circle marker",
        ref: circleRef,
        opacity: import_constants.MARKER_OPACITY,
        stroke: import_constants.MARKER_STROKE_COLOR,
        strokeWidth: import_constants.MARKER_STROKE_WIDTH,
        cx: x,
        cy: y,
        r: import_constants.MARKER_RADIUS,
        fill: color,
        onMouseMove,
        onMouseLeave,
        filter: hovered ? `url(#${brightnessFilterId})` : void 0
      }
    )
  ] });
};
