// @dynatrace/strato-components/charts/core/components/threshold-bar/markers/constants.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var constants_exports = {};
__export(constants_exports, {
  MARKER_DARK_BRIGHTNESS: () => MARKER_DARK_BRIGHTNESS,
  MARKER_LIGHT_BRIGHTNESS: () => MARKER_LIGHT_BRIGHTNESS,
  MARKER_OPACITY: () => MARKER_OPACITY,
  MARKER_RADIUS: () => MARKER_RADIUS,
  MARKER_SHADOW_OPACITY: () => MARKER_SHADOW_OPACITY,
  MARKER_STROKE_COLOR: () => MARKER_STROKE_COLOR,
  MARKER_STROKE_WIDTH: () => MARKER_STROKE_WIDTH
});
module.exports = __toCommonJS(constants_exports);
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
const MARKER_OPACITY = 0.85;
const MARKER_SHADOW_OPACITY = 0.5;
const MARKER_LIGHT_BRIGHTNESS = 0.8;
const MARKER_DARK_BRIGHTNESS = 1.3;
const MARKER_STROKE_COLOR = import_colors.default.Background.Base.Default;
const MARKER_STROKE_WIDTH = 1;
const MARKER_RADIUS = 4;
