// @dynatrace/strato-components/charts/core/components/threshold-bar/markers/PillMarker.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var PillMarker_exports = {};
__export(PillMarker_exports, {
  PillMarker: () => PillMarker
});
module.exports = __toCommonJS(PillMarker_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_group = require("@visx/group");
var import_react = require("react");
var import_constants = require("./constants.js");
var import_useId = require("../../../../../core/hooks/useId.js");
var import_useMouseHover = require("../../../hooks/useMouseHover.js");
var import_ThresholdBrightnessFilter = require("../ThresholdBrightnessFilter.js");
const PillMarker = (props) => {
  const { x, y0, y1, color, onMouseMove, onMouseLeave, opacity } = props;
  const rectRef = (0, import_react.useRef)(null);
  const hovered = (0, import_useMouseHover.useMouseHover)(rectRef);
  const id = (0, import_useId.useId)();
  const brightnessFilterId = `threshold-pill-marker-filter-${id}`;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_group.Group, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ThresholdBrightnessFilter.ThresholdBrightnessFilter, { id: brightnessFilterId }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "rect",
      {
        role: "graphics-symbol",
        "aria-label": "pill marker",
        ref: rectRef,
        x,
        y: Math.abs(y1 - y0) > import_constants.MARKER_RADIUS * 2 ? Math.min(y0, y1) : (y0 + y1) / 2 - import_constants.MARKER_RADIUS,
        height: Math.max(Math.abs(y1 - y0), import_constants.MARKER_RADIUS * 2),
        width: import_constants.MARKER_RADIUS * 2,
        fill: color,
        rx: import_constants.MARKER_RADIUS,
        opacity,
        filter: hovered ? `url(#${brightnessFilterId})` : void 0,
        stroke: import_constants.MARKER_STROKE_COLOR,
        strokeWidth: import_constants.MARKER_STROKE_WIDTH,
        onMouseMove,
        onMouseLeave
      }
    )
  ] });
};
