// @dynatrace/strato-components/tables/DataTable/tanstack-table-core-override/src/utils/getGroupedRowModel.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var getGroupedRowModel_exports = {};
__export(getGroupedRowModel_exports, {
  getGroupedRowModel: () => getGroupedRowModel
});
module.exports = __toCommonJS(getGroupedRowModel_exports);
var import_row = require("../core/row");
var import_utils = require("../utils");
function getGroupedRowModel() {
  return (table) => (0, import_utils.memo)(
    () => [table.getState().grouping, table.getPreGroupedRowModel()],
    (grouping, rowModel) => {
      if (!rowModel.rows.length || !grouping.length) {
        rowModel.rows.forEach((row) => {
          row.depth = 0;
          row.parentId = void 0;
        });
        return rowModel;
      }
      const existingGrouping = grouping.filter(
        (columnId) => table.getColumn(columnId)
      );
      const groupedFlatRows = [];
      const groupedRowsById = {};
      const groupUpRecursively = (rows, depth = 0, parentId) => {
        if (depth >= existingGrouping.length) {
          return rows.map((row) => {
            row.depth = depth;
            groupedFlatRows.push(row);
            groupedRowsById[row.id] = row;
            if (row.subRows) {
              row.subRows = groupUpRecursively(
                row.subRows,
                depth + 1,
                row.id
              );
            }
            return row;
          });
        }
        const columnId = existingGrouping[depth];
        const rowGroupsMap = groupBy(rows, columnId);
        const aggregatedGroupedRows = Array.from(rowGroupsMap.entries()).map(
          ([groupingValue, groupedRows2], index) => {
            let id = `${columnId}:${groupingValue}`;
            id = parentId ? `${parentId}>${id}` : id;
            const subRows = groupUpRecursively(groupedRows2, depth + 1, id);
            subRows.forEach((subRow) => {
              subRow.parentId = id;
            });
            const leafRows = depth ? (0, import_utils.flattenBy)(groupedRows2, (row2) => row2.subRows) : groupedRows2;
            const row = (0, import_row.createRow)(
              table,
              id,
              leafRows[0].original,
              index,
              depth,
              void 0,
              parentId
            );
            Object.assign(row, {
              groupingColumnId: columnId,
              groupingValue,
              subRows,
              leafRows,
              getValue: (columnId2) => {
                if (existingGrouping.includes(columnId2)) {
                  if (row._valuesCache.hasOwnProperty(columnId2)) {
                    return row._valuesCache[columnId2];
                  }
                  if (groupedRows2[0]) {
                    row._valuesCache[columnId2] = groupedRows2[0].getValue(columnId2) ?? void 0;
                  }
                  return row._valuesCache[columnId2];
                }
                if (row._groupingValuesCache.hasOwnProperty(columnId2)) {
                  return row._groupingValuesCache[columnId2];
                }
                const column = table.getColumn(columnId2);
                const aggregateFn = column?.getAggregationFn();
                if (aggregateFn) {
                  row._groupingValuesCache[columnId2] = aggregateFn(
                    columnId2,
                    leafRows,
                    groupedRows2
                  );
                  return row._groupingValuesCache[columnId2];
                }
              }
            });
            subRows.forEach((subRow) => {
              groupedFlatRows.push(subRow);
              groupedRowsById[subRow.id] = subRow;
            });
            return row;
          }
        );
        return aggregatedGroupedRows;
      };
      const groupedRows = groupUpRecursively(rowModel.rows, 0);
      groupedRows.forEach((subRow) => {
        groupedFlatRows.push(subRow);
        groupedRowsById[subRow.id] = subRow;
      });
      return {
        rows: groupedRows,
        flatRows: groupedFlatRows,
        rowsById: groupedRowsById
      };
    },
    (0, import_utils.getMemoOptions)(table.options, "debugTable", "getGroupedRowModel", () => {
      table._queue(() => {
        table._autoResetExpanded();
        table._autoResetPageIndex();
      });
    })
  );
}
function groupBy(rows, columnId) {
  const groupMap = /* @__PURE__ */ new Map();
  return rows.reduce((map, row) => {
    const resKey = `${row.getGroupingValue(columnId)}`;
    const previous = map.get(resKey);
    if (!previous) {
      map.set(resKey, [row]);
    } else {
      previous.push(row);
    }
    return map;
  }, groupMap);
}
