// @dynatrace/strato-components/charts/core/components/overlay-tooltip/constants.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var constants_exports = {};
__export(constants_exports, {
  DEFAULT_HEADER_LINES: () => DEFAULT_HEADER_LINES,
  DEFAULT_SERIES_NAME_LINES: () => DEFAULT_SERIES_NAME_LINES,
  DEFAULT_TRUNCATION_MODE: () => DEFAULT_TRUNCATION_MODE,
  GAP_BETWEEN_COLUMNS: () => GAP_BETWEEN_COLUMNS,
  MAX_HEADER_LINES: () => MAX_HEADER_LINES,
  MAX_SERIES_NAME_LINES: () => MAX_SERIES_NAME_LINES,
  MAX_TOOLTIP_WIDTH: () => MAX_TOOLTIP_WIDTH,
  MIN_TOOLTIP_WIDTH: () => MIN_TOOLTIP_WIDTH,
  SCROLLBAR_WIDTH: () => SCROLLBAR_WIDTH,
  SPACE_CHARACTER_WIDTH: () => SPACE_CHARACTER_WIDTH,
  SYMBOL_AND_SPACING_WIDTH: () => SYMBOL_AND_SPACING_WIDTH,
  TOOLTIP_CONTENT_PADDING: () => TOOLTIP_CONTENT_PADDING,
  TOOLTIP_HEADER_PADDING: () => TOOLTIP_HEADER_PADDING,
  TOOLTIP_WRAPPER_PADDING: () => TOOLTIP_WRAPPER_PADDING
});
module.exports = __toCommonJS(constants_exports);
const TOOLTIP_HEADER_PADDING = 6;
const TOOLTIP_WRAPPER_PADDING = 4;
const TOOLTIP_CONTENT_PADDING = TOOLTIP_HEADER_PADDING + TOOLTIP_WRAPPER_PADDING;
const DEFAULT_HEADER_LINES = 3;
const MAX_HEADER_LINES = 5;
const DEFAULT_SERIES_NAME_LINES = 1;
const MAX_SERIES_NAME_LINES = 2;
const DEFAULT_TRUNCATION_MODE = "end";
const SPACE_CHARACTER_WIDTH = 4;
const SYMBOL_CONTAINER_WIDTH = 24;
const GAP_BETWEEN_COLUMNS = 12;
const SYMBOL_AND_SPACING_WIDTH = SYMBOL_CONTAINER_WIDTH + GAP_BETWEEN_COLUMNS;
const MIN_TOOLTIP_WIDTH = 200;
const MAX_TOOLTIP_WIDTH = 480;
const SCROLLBAR_WIDTH = 14;
