// @dynatrace/strato-components/charts/core/components/overlay-tooltip/defaults.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var defaults_exports = {};
__export(defaults_exports, {
  DEFAULTS: () => DEFAULTS,
  GEO_DEFAULTS: () => GEO_DEFAULTS
});
module.exports = __toCommonJS(defaults_exports);
var import_format_value = require("./components/header/utils/format-value.js");
var import_value_formatters = require("./components/header/utils/value-formatters.js");
var import_messages = require("./messages.js");
const BASE_CONFIG = {
  defaultActionsConfig: /* @__PURE__ */ new Map([
    [
      "name",
      {
        field: "seriesLabel"
      }
    ],
    [
      "value",
      {
        field: "valueLabel"
      }
    ]
  ])
};
const BASE_COORDINATES_CONFIG = {
  defaultActionsConfig: /* @__PURE__ */ new Map([
    [
      "name",
      {
        field: "valueLabel"
      }
    ]
  ])
};
const BASE_XY_CONFIG = {
  ...BASE_CONFIG,
  headerFormatStrategy: import_format_value.dotFormatterStrategy,
  headerLabel: import_messages.OVERLAY_TOOLTIP_MESSAGES.xValue,
  seriesLabel: import_messages.OVERLAY_TOOLTIP_MESSAGES.series,
  valueLabel: import_messages.OVERLAY_TOOLTIP_MESSAGES.yValue,
  sharedVariant: true
};
const GEO_DEFAULTS = {
  bubble: {
    ...BASE_COORDINATES_CONFIG,
    seriesLabel: import_messages.OVERLAY_TOOLTIP_MESSAGES.location,
    valueLabel: import_messages.OVERLAY_TOOLTIP_MESSAGES.coordinates,
    sharedVariant: true
  },
  choropleth: {
    defaultActionsConfig: /* @__PURE__ */ new Map([
      [
        "name",
        {
          field: "seriesLabel"
        }
      ]
    ]),
    seriesLabel: import_messages.OVERLAY_TOOLTIP_MESSAGES.location,
    valueLabel: import_messages.OVERLAY_TOOLTIP_MESSAGES.value
  },
  connection: {
    ...BASE_COORDINATES_CONFIG,
    seriesLabel: import_messages.OVERLAY_TOOLTIP_MESSAGES.location,
    valueLabel: import_messages.OVERLAY_TOOLTIP_MESSAGES.coordinates,
    sharedVariant: true
  },
  dot: {
    ...BASE_COORDINATES_CONFIG,
    seriesLabel: import_messages.OVERLAY_TOOLTIP_MESSAGES.location,
    valueLabel: import_messages.OVERLAY_TOOLTIP_MESSAGES.coordinates
  }
};
const DEFAULTS = {
  dot: { ...BASE_XY_CONFIG, sharedVariant: false },
  line: { ...BASE_XY_CONFIG },
  area: { ...BASE_XY_CONFIG },
  bar: { ...BASE_XY_CONFIG },
  rect: {
    ...BASE_CONFIG,
    symbolPosition: "right",
    headerFormatStrategy: import_value_formatters.formatValueRange,
    headerLabel: import_messages.OVERLAY_TOOLTIP_MESSAGES.xValue,
    seriesLabel: import_messages.OVERLAY_TOOLTIP_MESSAGES.yValue,
    valueLabel: import_messages.OVERLAY_TOOLTIP_MESSAGES.value,
    headerShowRangeDuration: true
  },
  pie: {
    ...BASE_CONFIG,
    headerFormatStrategy: import_value_formatters.emptyHeaderFormatter,
    seriesLabel: import_messages.OVERLAY_TOOLTIP_MESSAGES.category,
    valueLabel: import_messages.OVERLAY_TOOLTIP_MESSAGES.value
  },
  meterbar: {
    ...BASE_CONFIG,
    seriesLabel: import_messages.OVERLAY_TOOLTIP_MESSAGES.segment,
    valueLabel: import_messages.OVERLAY_TOOLTIP_MESSAGES.value
  },
  multimeterbar: {
    ...BASE_CONFIG,
    seriesLabel: import_messages.OVERLAY_TOOLTIP_MESSAGES.segment,
    valueLabel: import_messages.OVERLAY_TOOLTIP_MESSAGES.value,
    sharedVariant: true
  },
  treemap: {
    ...BASE_CONFIG,
    seriesLabel: import_messages.OVERLAY_TOOLTIP_MESSAGES.node,
    valueLabel: import_messages.OVERLAY_TOOLTIP_MESSAGES.value
  },
  honeycomb: {
    ...BASE_CONFIG,
    headerFormatStrategy: import_value_formatters.emptyHeaderFormatter,
    symbolPosition: "right",
    seriesLabel: import_messages.OVERLAY_TOOLTIP_MESSAGES.node,
    valueLabel: import_messages.OVERLAY_TOOLTIP_MESSAGES.value
  },
  geoBubble: GEO_DEFAULTS.bubble,
  geoChoropleth: GEO_DEFAULTS.choropleth,
  geoConnection: GEO_DEFAULTS.connection,
  geoDot: GEO_DEFAULTS.dot
};
