// @dynatrace/strato-components/charts/core/components/overlay-tooltip/overlay.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var overlay_exports = {};
__export(overlay_exports, {
  buildOverlay: () => buildOverlay
});
module.exports = __toCommonJS(overlay_exports);
const NullPositions = {
  absoluteX: null,
  absoluteY: null,
  relativeX: null,
  relativeY: null,
  percentX: null,
  percentY: null
};
function getAdjustedPositionsFromAbsolute(absoluteX, absoluteY) {
  const bounds = this.getBounds();
  if (!bounds) {
    return NullPositions;
  }
  const { left, top, width, height } = bounds;
  const relativeX = absoluteX - left;
  const relativeY = absoluteY - top;
  const percentX = relativeX * 100 / width;
  const percentY = relativeY * 100 / height;
  return {
    absoluteX,
    absoluteY,
    relativeX,
    relativeY,
    percentX,
    percentY
  };
}
function getAdjustedPositionsFromPercent(percentX, percentY, relativeX, relativeY) {
  const bounds = this.getBounds();
  if (!bounds) {
    return NullPositions;
  }
  const { left, top, width, height } = bounds;
  relativeX ||= percentX * width / 100;
  relativeY ||= percentY * height / 100;
  const absoluteX = Math.round(left + relativeX);
  const absoluteY = Math.round(top + relativeY);
  return {
    absoluteX,
    absoluteY,
    relativeX,
    relativeY,
    percentX,
    percentY
  };
}
function getAdjustedPositionsFromRelative(relativeX, relativeY) {
  const bounds = this.getBounds();
  if (!bounds) {
    return NullPositions;
  }
  const { left, top, width, height } = bounds;
  const percentX = relativeX * 100 / width;
  const percentY = relativeY * 100 / height;
  const absoluteX = Math.round(left + relativeX);
  const absoluteY = Math.round(top + relativeY);
  return {
    absoluteX,
    absoluteY,
    relativeX,
    relativeY,
    percentX,
    percentY
  };
}
function buildOverlay(overlayRef) {
  let bounds = void 0;
  return {
    overlayRef,
    getBounds() {
      bounds ||= overlayRef.current?.getBoundingClientRect();
      return bounds;
    },
    getAdjustedPositionsFromAbsolute,
    getAdjustedPositionsFromPercent,
    getAdjustedPositionsFromRelative,
    clear() {
      bounds = void 0;
    }
  };
}
