// @dynatrace/strato-components/charts/core/components/overlay-tooltip/OverlayTooltip.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var OverlayTooltip_exports = {};
__export(OverlayTooltip_exports, {
  ChartTooltip: () => ChartTooltip,
  OverlayTooltip: () => OverlayTooltip,
  Tooltip: () => Tooltip
});
module.exports = __toCommonJS(OverlayTooltip_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_DynamicTooltipRenderer = require("./components/DynamicTooltipRenderer.js");
var import_Footer = require("./components/Footer.js");
var import_Action = require("./components/global-actions/Action.js");
var import_ActionsPlaceholder = require("./components/global-actions/ActionsPlaceholder.js");
var import_Header = require("./components/header/Header.js");
var import_Series = require("./components/series/Series.js");
var import_SeriesHeader = require("./components/series/SeriesHeader.js");
var import_SeriesItem = require("./components/series/SeriesItem.js");
var import_TooltipPortal = require("./components/TooltipPortal.js");
const TooltipRoot = (props) => {
  if (props.absoluteX !== void 0 && props.absoluteY !== void 0) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_TooltipPortal.TooltipPortal,
      {
        isInteractive: props.isInteractive ?? false,
        absoluteX: props.absoluteX,
        absoluteY: props.absoluteY,
        children: props.children
      }
    );
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_DynamicTooltipRenderer.DynamicTooltipRenderer, { ...props });
};
const OverlayTooltip = Object.assign(TooltipRoot, {
  Action: import_Action.Action,
  ActionsPlaceholder: import_ActionsPlaceholder.ActionsPlaceholder,
  Footer: import_Footer.Footer,
  Header: import_Header.Header,
  Series: import_Series.Series,
  SeriesHeader: import_SeriesHeader.SeriesHeader,
  SeriesItem: import_SeriesItem.SeriesItem
});
const Tooltip = () => null;
Tooltip.displayName = "ChartTooltip";
const ChartTooltip = Object.assign(Tooltip, {
  Action: import_Action.Action,
  ActionsPlaceholder: import_ActionsPlaceholder.ActionsPlaceholder,
  Footer: import_Footer.Footer,
  Header: import_Header.Header,
  Series: import_Series.Series,
  SeriesHeader: import_SeriesHeader.SeriesHeader,
  SeriesItem: import_SeriesItem.SeriesItem
});
