// @dynatrace/strato-components/charts/core/components/overlay-tooltip/types.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var types_exports = {};
__export(types_exports, {
  GeometryTypesEnum: () => GeometryTypesEnum
});
module.exports = __toCommonJS(types_exports);
var GeometryTypesEnum = /* @__PURE__ */ ((GeometryTypesEnum2) => {
  GeometryTypesEnum2["dot"] = "dot";
  GeometryTypesEnum2["line"] = "line";
  GeometryTypesEnum2["area"] = "area";
  GeometryTypesEnum2["bar"] = "bar";
  GeometryTypesEnum2["pie"] = "pie";
  GeometryTypesEnum2["meterbar"] = "meterbar";
  GeometryTypesEnum2["multimeterbar"] = "multimeterbar";
  GeometryTypesEnum2["treemap"] = "treemap";
  GeometryTypesEnum2["rect"] = "rect";
  GeometryTypesEnum2["honeycomb"] = "honeycomb";
  GeometryTypesEnum2["geoBubble"] = "geoBubble";
  GeometryTypesEnum2["geoChoropleth"] = "geoChoropleth";
  GeometryTypesEnum2["geoConnection"] = "geoConnection";
  GeometryTypesEnum2["geoDot"] = "geoDot";
  return GeometryTypesEnum2;
})(GeometryTypesEnum || {});
