// @dynatrace/strato-components/charts/core/components/overlay-tooltip/useTooltipReducer.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useTooltipReducer_exports = {};
__export(useTooltipReducer_exports, {
  useTooltipReducer: () => useTooltipReducer
});
module.exports = __toCommonJS(useTooltipReducer_exports);
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_OverlayContext = require("./contexts/OverlayContext.js");
var import_TooltipContext = require("./contexts/TooltipContext.js");
var import_utils = require("./utils.js");
const useTooltipReducer = () => {
  const store = (0, import_TooltipContext.useTooltip)();
  const overlay = (0, import_OverlayContext.useOverlay)();
  const dispatch = (0, import_react.useCallback)(
    (action) => () => {
      switch (action.type) {
        case "RESET_TOOLTIP":
          overlay.clear();
          store.setState(() => ({
            pinned: false,
            absoluteX: null,
            absoluteY: null,
            relativeX: null,
            relativeY: null,
            percentX: null,
            percentY: null,
            inBounds: false,
            activeSeries: null,
            data: (0, import_utils.emptyData)()
          }));
          break;
        case "SET_PINNED":
          store.setState((prevState) => {
            prevState.inBounds || overlay.clear();
            const { pinned, data, absoluteX, absoluteY } = action.payload;
            const activeSeries = pinned ? prevState.activeSeries : null;
            if ((0, import_lodash_es.isNil)(absoluteX) || (0, import_lodash_es.isNil)(absoluteY)) {
              return {
                ...prevState,
                inBounds: true,
                pinned,
                data,
                activeSeries
              };
            }
            const positions = overlay.getAdjustedPositionsFromAbsolute(
              absoluteX,
              absoluteY
            );
            return {
              ...prevState,
              pinned,
              inBounds: true,
              data,
              activeSeries,
              ...positions
            };
          });
          break;
        case "SET_POSITION":
          store.setState((prevState) => {
            if (prevState.pinned) {
              return prevState;
            }
            prevState.inBounds || overlay.clear();
            const { absoluteX, absoluteY, data } = action.payload;
            const positions = overlay.getAdjustedPositionsFromAbsolute(
              absoluteX,
              absoluteY
            );
            return {
              ...prevState,
              inBounds: true,
              ...positions,
              data: data || (0, import_utils.emptyData)()
            };
          });
          break;
        case "SET_OUT_BOUNDS":
          store.setState((prevState) => {
            overlay.clear();
            if (prevState.pinned) {
              return prevState;
            }
            return {
              ...prevState,
              absoluteX: null,
              absoluteY: null,
              relativeX: null,
              relativeY: null,
              percentX: null,
              percentY: null,
              inBounds: false,
              data: (0, import_utils.emptyData)()
            };
          });
          break;
        case "UPDATE_POSITION":
          store.setState((prevState) => {
            overlay.clear();
            const { pinned, percentX, percentY, relativeX, relativeY } = prevState;
            if (!pinned || (0, import_lodash_es.isNil)(percentX) || (0, import_lodash_es.isNil)(percentY)) {
              return prevState;
            }
            const positions = overlay.getAdjustedPositionsFromPercent(
              percentX,
              percentY,
              ...action.payload.type === "non-resize" ? [relativeX, relativeY] : [void 0, void 0]
            );
            return {
              ...prevState,
              ...positions,
              activeSeries: action.payload.scroll ? null : prevState.activeSeries
            };
          });
          break;
        case "SET_ACTIVE_SERIES":
          store.setState((prevState) => {
            if (prevState.activeSeries === action.payload.activeSeries) {
              return prevState;
            }
            return {
              ...prevState,
              activeSeries: action.payload.activeSeries
            };
          });
          break;
        case "TOOLTIP_MOUSE_LEAVE":
          store.setState((prevState) => {
            if ((0, import_lodash_es.isNull)(prevState.activeSeries)) {
              return prevState;
            }
            return {
              ...prevState,
              activeSeries: null
            };
          });
          break;
        default:
          throw new Error("Unhandled action type");
      }
    },
    [store, overlay]
  );
  return dispatch;
};
