// @dynatrace/strato-components/charts/core/components/overlay-tooltip/useTooltipActions.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useTooltipActions_exports = {};
__export(useTooltipActions_exports, {
  useTooltipActions: () => useTooltipActions
});
module.exports = __toCommonJS(useTooltipActions_exports);
var import_react = require("react");
var import_TooltipContext = require("./contexts/TooltipContext.js");
var import_useTooltipReducer = require("./useTooltipReducer.js");
var import_utils = require("./utils.js");
const useTooltipActions = () => {
  const dispatch = (0, import_useTooltipReducer.useTooltipReducer)();
  const tooltipState = (0, import_TooltipContext.useTooltip)();
  const hover = (0, import_react.useCallback)(
    (event, data) => {
      dispatch({
        type: "SET_POSITION",
        payload: {
          absoluteX: Math.round(event.clientX),
          absoluteY: Math.round(event.clientY),
          data
        }
      })();
    },
    [dispatch]
  );
  const pin = (0, import_react.useCallback)(
    (event, data) => {
      const { absoluteX, absoluteY } = tooltipState.getState();
      const coords = absoluteX !== null && absoluteY !== null ? { clientX: absoluteX, clientY: absoluteY } : {
        clientX: Math.round(event.clientX),
        clientY: Math.round(event.clientY)
      };
      dispatch({
        type: "SET_PINNED",
        payload: {
          pinned: true,
          absoluteX: coords.clientX,
          absoluteY: coords.clientY,
          data
        }
      })();
    },
    [dispatch, tooltipState]
  );
  const unpin = (0, import_react.useCallback)(() => {
    dispatch({
      type: "SET_PINNED",
      payload: { pinned: false, data: (0, import_utils.emptyData)() }
    })();
  }, [dispatch]);
  const togglePin = (0, import_react.useCallback)(
    (event, data) => {
      if (tooltipState.getState().pinned) {
        unpin();
      } else {
        pin(event, data);
      }
    },
    [pin, unpin, tooltipState]
  );
  const reset = (0, import_react.useCallback)(() => {
    dispatch({ type: "RESET_TOOLTIP" })();
  }, [dispatch]);
  const outOfBounds = (0, import_react.useCallback)(() => {
    dispatch({ type: "SET_OUT_BOUNDS" })();
  }, [dispatch]);
  return { hover, pin, togglePin, unpin, reset, outOfBounds };
};
