// @dynatrace/strato-components/esm/charts/gauge/components/GaugeLayoutOverlay/GaugeLayoutOverlay.js
import { jsx, jsxs } from "react/jsx-runtime";
import {
  container,
  tooltipOverlay,
  tooltipOverlayFull,
  consumerLoadingStyle
} from "./GaugeLayoutOverlay.sty.js";
import { Flex } from "../../../../layouts/flex/Flex.js";
import { LoaderOverlay } from "../../../core/components/loader-overlay/LoaderOverlay.js";
import { useOverlay } from "../../../core/components/overlay-tooltip/contexts/OverlayContext.js";
import { useTooltipActions } from "../../../core/components/overlay-tooltip/useTooltipActions.js";
import { useGaugeChartTooltip } from "../../contexts/GaugeChartTooltipContext.js";
import { useGaugeChartTooltipOverlayBounds } from "../../hooks/useGaugeChartTooltipOverlayBounds.js";
import {
  GaugeChartToolbar
} from "../GaugeChartLayout/GaugeChartToolbar.js";
import { GaugeOverlayTooltip } from "../GaugeChartTooltip/GaugeOverlayTooltip.js";
const GaugeChartOverlay = (props) => {
  const { children, value, color, intents, toolbar, showLoading } = props;
  const overlay = useOverlay();
  const overlayBounds = useGaugeChartTooltipOverlayBounds();
  const { name: tooltipName, hidden: tooltipHidden } = useGaugeChartTooltip();
  const { hover, togglePin, outOfBounds } = useTooltipActions();
  const data = {
    geometry: "meterbar",
    series: [{ name: tooltipName ?? "Value", value, color }]
  };
  const overlayClassName = overlayBounds ? tooltipOverlay : `${tooltipOverlay} ${tooltipOverlayFull}`;
  const overlayStyle = overlayBounds ? {
    top: overlayBounds.top,
    left: overlayBounds.left,
    width: overlayBounds.width,
    height: overlayBounds.height
  } : void 0;
  return /* @__PURE__ */ jsxs(
    Flex,
    {
      ref: overlay.overlayRef,
      "data-testid": "gauge-overlay-wrapper",
      className: container({ showLoading }),
      children: [
        /* @__PURE__ */ jsx(GaugeChartToolbar, { intents, toolbar }),
        !showLoading && /* @__PURE__ */ jsx(
          Flex,
          {
            className: overlayClassName,
            onPointerMove: (event) => hover(event, data),
            onPointerLeave: outOfBounds,
            onClick: (event) => togglePin(event, data),
            style: overlayStyle,
            children: !tooltipHidden && /* @__PURE__ */ jsx(GaugeOverlayTooltip, { value, color })
          }
        ),
        children,
        showLoading ? /* @__PURE__ */ jsx(
          LoaderOverlay,
          {
            gridArea: "1/1/2/2",
            consumerStyle: consumerLoadingStyle
          }
        ) : null
      ]
    }
  );
};
GaugeChartOverlay["displayName"] = "GaugeChartOverlay";
export {
  GaugeChartOverlay
};
