// @dynatrace/strato-components/esm/charts/histogram/components/axes/HistogramValueLeftAxisArea.js
import { jsx } from "react/jsx-runtime";
import { isEmpty } from "lodash-es";
import { ValueLeftAxis } from "../../../core/components/axes/components/ValueLeftAxis.js";
import { ChartGrid } from "../../../core/components/chart-grid/ChartGridRoot.js";
import { AREAS } from "../../constants/histogram-grid-area-rules.js";
import { useHistogramTickAxis } from "../../hooks/useHistogramTickAxis.js";
import { useHistogramYScales } from "../../hooks/useHistogramYScales.js";
import { useYAxisConfig } from "../../hooks/useYAxisConfig.js";
const HistogramValueLeftAxisArea = () => {
  const { left: axisLeft } = useYAxisConfig();
  const ticks = useHistogramTickAxis();
  const [yScale] = useHistogramYScales();
  const leftTicks = ticks?.left || { ticks: [], formattedTicks: [] };
  if (isEmpty(leftTicks.ticks) || !yScale) {
    return null;
  }
  return /* @__PURE__ */ jsx(ChartGrid.GridArea, { gridArea: AREAS.leftAxis, children: ({ availableWidth, availableHeight }) => {
    return /* @__PURE__ */ jsx(
      ValueLeftAxis,
      {
        ticks: leftTicks.ticks,
        formattedTicks: leftTicks.formattedTicks,
        containerHeight: availableHeight,
        containerWidth: availableWidth,
        scale: yScale,
        label: axisLeft?.label
      }
    );
  } });
};
HistogramValueLeftAxisArea["displayName"] = "HistogramValueLeftAxisArea";
export {
  HistogramValueLeftAxisArea
};
