// @dynatrace/strato-components/charts/core/components/axes/components/utils/getCategoryBottomAxisProps.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var getCategoryBottomAxisProps_exports = {};
__export(getCategoryBottomAxisProps_exports, {
  getCategoryBottomAxisProps: () => getCategoryBottomAxisProps
});
module.exports = __toCommonJS(getCategoryBottomAxisProps_exports);
var import_constants = require("../../constants.js");
const VISUAL_ADJUSTMENT = 2;
const getCategoryBottomAxisProps = (renderX, isTickRotated, maxSize, tickNameHasEmoji) => {
  const TICK_LINE_HEIGHT_VALIDATED = tickNameHasEmoji ? import_constants.TICK_LINE_HEIGHT + import_constants.TICK_LINE_ICON_OFFSET : import_constants.TICK_LINE_HEIGHT;
  const x = isTickRotated ? -(maxSize + VISUAL_ADJUSTMENT * 2) : renderX - maxSize / VISUAL_ADJUSTMENT;
  const y = isTickRotated ? renderX - TICK_LINE_HEIGHT_VALIDATED / 2 : VISUAL_ADJUSTMENT;
  const foreignObjectInlineStyle = {
    height: TICK_LINE_HEIGHT_VALIDATED,
    width: maxSize,
    transform: isTickRotated ? "rotate(-90deg)" : "none"
  };
  const foreignObjectContentInlineStyle = {
    ...import_constants.axisFontProps,
    textAlign: isTickRotated ? "right" : "center",
    height: TICK_LINE_HEIGHT_VALIDATED,
    width: maxSize,
    whiteSpace: "nowrap"
  };
  return { x, y, foreignObjectInlineStyle, foreignObjectContentInlineStyle };
};
