// @dynatrace/strato-components/charts/categorical-bar/utils/categorical-config.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var categorical_config_exports = {};
__export(categorical_config_exports, {
  buildCategoricalBarChartConfig: () => buildCategoricalBarChartConfig,
  deserializeConfig: () => deserializeConfig,
  getColorPaletteConfiguration: () => getColorPaletteConfiguration,
  getColorPaletteMode: () => getColorPaletteMode,
  getGroupModeConfiguration: () => getGroupModeConfiguration,
  getLayoutConfiguration: () => getLayoutConfiguration,
  getThresholdsConfig: () => getThresholdsConfig,
  getValueRepresentationConfiguration: () => getValueRepresentationConfiguration
});
module.exports = __toCommonJS(categorical_config_exports);
var import_lodash_es = require("lodash");
var import_is_valid_group_mode = require("./is-valid-group-mode.js");
var import_is_valid_tooltip_variant = require("./is-valid-tooltip-variant.js");
var import_ColorRule = require("../../core/slots/ColorRule.js");
var import_color_palette = require("../../core/types/color-palette.js");
var import_getValidColor = require("../../core/utils/getValidColor.js");
var import_is_valid_color_palette = require("../../core/utils/is-valid-color-palette.js");
var import_is_valid_legend_position = require("../../core/utils/is-valid-legend-position.js");
var import_is_valid_value_representation = require("../../core/utils/is-valid-value-representation.js");
var import_constants = require("../constants.js");
var import_defaults = require("../types/defaults.js");
var import_is_valid_layout = require("../utils/is-valid-layout.js");
function getColorPaletteConfiguration(contextColorPalette, propsColorPalette) {
  const colorPalette = propsColorPalette ?? contextColorPalette ?? import_defaults.defaultConfig.colorPalette;
  return (0, import_color_palette.isCustomColorPalette)(colorPalette) || (0, import_is_valid_color_palette.isValidColorPalette)(colorPalette) ? colorPalette : import_defaults.defaultConfig.colorPalette;
}
function getColorPaletteMode(propsColorPaletteMode, contextColorPaletteMode) {
  const colorPaletteMode = propsColorPaletteMode ?? contextColorPaletteMode ?? import_defaults.defaultConfig.colorPaletteMode;
  return colorPaletteMode;
}
function getLayoutConfiguration(contextLayout, propsLayout) {
  const layout = propsLayout ?? contextLayout ?? import_defaults.defaultConfig.layout;
  return (0, import_is_valid_layout.isValidLayout)(layout) ? layout : import_defaults.defaultConfig.layout;
}
function getGroupModeConfiguration(contextGroupMode, propsGroupMode) {
  const groupMode = propsGroupMode ?? contextGroupMode ?? import_defaults.defaultConfig.groupMode;
  return (0, import_is_valid_group_mode.isValidGroupMode)(groupMode) ? groupMode : import_defaults.defaultConfig.groupMode;
}
function getValueRepresentationConfiguration(contextValueRepresentation, propsValueRepresentation) {
  const valueRepresentation = propsValueRepresentation ?? contextValueRepresentation ?? import_defaults.defaultConfig.valueRepresentation;
  return (0, import_is_valid_value_representation.isValidValueRepresentation)(valueRepresentation) ? valueRepresentation : import_defaults.defaultConfig.valueRepresentation;
}
function getThresholdsConfig(contextThresholds = [], slotsThresholds) {
  return contextThresholds.concat(slotsThresholds).map((threshold) => ({
    ...threshold,
    id: crypto.randomUUID(),
    color: (0, import_getValidColor.isValidCSSColor)(threshold.color) ? threshold.color : import_constants.DEFAULT_SCALE_FALLBACK_COLOR
  }));
}
function buildCategoricalBarChartConfig(props, contextConfig, slots) {
  const colorPalette = getColorPaletteConfiguration(
    contextConfig.colorPalette,
    props.colorPalette
  );
  const colorPaletteMode = getColorPaletteMode(
    props.colorPaletteMode,
    contextConfig.colorPaletteMode
  );
  const layout = getLayoutConfiguration(contextConfig.layout, props.layout);
  const groupMode = getGroupModeConfiguration(
    contextConfig.groupMode,
    props.groupMode
  );
  const valueRepresentation = getValueRepresentationConfiguration(
    contextConfig.valueRepresentation,
    props.valueRepresentation
  );
  const categoryAxis = (0, import_lodash_es.merge)(
    {},
    import_defaults.defaultConfig.categoryAxis,
    contextConfig.categoryAxis,
    slots.categoryAxis
  );
  categoryAxis.tickLabelLayout = (0, import_is_valid_layout.isValidLayout)(categoryAxis.tickLabelLayout) ? categoryAxis.tickLabelLayout : import_defaults.defaultConfig.categoryAxis.tickLabelLayout;
  const valueAxis = (0, import_lodash_es.merge)(
    {},
    import_defaults.defaultConfig.valueAxis,
    contextConfig.valueAxis,
    slots.valueAxis
  );
  const legend = (0, import_lodash_es.merge)(
    {},
    import_defaults.defaultConfig.legend,
    contextConfig.legend,
    slots.legend
  );
  legend.position = (0, import_is_valid_legend_position.isValidLegendPosition)(legend.position) ? legend.position : import_defaults.defaultConfig.legend.position;
  legend.hidden = Boolean(legend.hidden);
  const tooltip = (0, import_lodash_es.merge)(
    {},
    import_defaults.defaultConfig.tooltip,
    contextConfig.tooltip,
    slots.tooltip
  );
  tooltip.variant = (0, import_is_valid_tooltip_variant.isValidTooltipVariant)(tooltip.variant) ? tooltip.variant : import_defaults.defaultConfig.tooltip.variant;
  const toolbar = (0, import_lodash_es.merge)(
    {},
    import_defaults.defaultConfig.toolbar,
    contextConfig.toolbar,
    slots.toolbar
  );
  const truncationMode = props.truncationMode ?? contextConfig.truncationMode ?? import_defaults.defaultConfig.truncationMode;
  const configThresholds = contextConfig.thresholdIndicators || [];
  const thresholdIndicators = getThresholdsConfig(
    configThresholds,
    slots.thresholds
  );
  const colorRules = (contextConfig.colorRules ?? []).filter(import_ColorRule.isValidColorRule).concat(slots.colorRules ?? []);
  return {
    colorPalette,
    colorPaletteMode,
    layout,
    groupMode,
    valueRepresentation,
    valueAxis,
    categoryAxis,
    legend,
    tooltip,
    toolbar,
    truncationMode,
    thresholdIndicators,
    colorRules,
    intents: slots.intents
  };
}
function deserializeConfig(serializedConfig) {
  try {
    return (0, import_lodash_es.merge)({}, import_defaults.defaultConfig, JSON.parse(serializedConfig));
  } catch {
    return import_defaults.defaultConfig;
  }
}
