// @dynatrace/strato-components/charts/categorical-bar/utils/categorical-layout.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var categorical_layout_exports = {};
__export(categorical_layout_exports, {
  getBarCoordinates: () => getBarCoordinates,
  getBarSize: () => getBarSize,
  getTopValueCoord: () => getTopValueCoord
});
module.exports = __toCommonJS(categorical_layout_exports);
var import_layout = require("./layout.js");
var import_safe_get = require("../../core/utils/safe-get.js");
function getTopValueCoord(dimension, valueScale, dimensionMap) {
  const { value, cumulativeValue } = (0, import_safe_get.safeGet)(dimensionMap, dimension);
  return valueScale((cumulativeValue ?? 0) + value) ?? 0;
}
function getBarSize(value, valueScale) {
  const zeroCoordinate = valueScale(0) ?? 0;
  const scaledValue = valueScale(value) ?? 0;
  if (value >= 0) {
    return zeroCoordinate - scaledValue;
  }
  return scaledValue - zeroCoordinate;
}
function getValueCoord(dimension, valueScale, dimensionMap) {
  const { value, cumulativeValue } = (0, import_safe_get.safeGet)(dimensionMap, dimension);
  return valueScale(
    value >= 0 ? (cumulativeValue ?? 0) + value : cumulativeValue ?? 0
  ) ?? 0;
}
function getBarCoordinates(groupMode, categoryScale, valueScale, dimensionsScale, isSingleDimension, dimensionMap, categoryMap) {
  if ((0, import_layout.isStacked)(groupMode) || isSingleDimension) {
    return {
      categoryCoord: (dimension, category) => categoryScale((0, import_safe_get.safeGet)(categoryMap, category).id) ?? 0,
      valueCoord: (dimension) => getValueCoord(dimension, valueScale, dimensionMap),
      barThickness: () => categoryScale.bandwidth(),
      barSize: (dimension) => getBarSize((0, import_safe_get.safeGet)(dimensionMap, dimension).value, valueScale)
    };
  }
  return {
    categoryCoord: (dimension, category) => {
      const categoryCoord = categoryScale((0, import_safe_get.safeGet)(categoryMap, category).id) ?? 0;
      const dimensionCoord = dimensionsScale((0, import_safe_get.safeGet)(dimensionMap, dimension).name) ?? 0;
      return categoryCoord + dimensionCoord;
    },
    valueCoord: (dimension) => getValueCoord(dimension, valueScale, dimensionMap),
    barThickness: () => dimensionsScale.bandwidth(),
    barSize: (dimension) => getBarSize((0, import_safe_get.safeGet)(dimensionMap, dimension).value, valueScale)
  };
}
