// @dynatrace/strato-components/charts/categorical-bar/utils/generate-categorical-bar-data-helpers.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var generate_categorical_bar_data_helpers_exports = {};
__export(generate_categorical_bar_data_helpers_exports, {
  getDimensionColor: () => getDimensionColor,
  handleMultipleDimensions: () => handleMultipleDimensions,
  handleSingleDimension: () => handleSingleDimension
});
module.exports = __toCommonJS(generate_categorical_bar_data_helpers_exports);
var import_lodash_es = require("lodash");
var import_color_palette = require("../../core/types/color-palette.js");
var import_get_color_palette = require("../../core/utils/get-color-palette.js");
var import_is_valid_color_palette = require("../../core/utils/is-valid-color-palette.js");
var import_constants = require("../constants.js");
function isColorToken(value) {
  return !(0, import_lodash_es.isNil)(value) && typeof value !== "string";
}
function isColorPalette(value) {
  return typeof value === "string";
}
function getColorRuleResult(hasColorRules, colorRules, seriesForColorRule, categoryData, dimensionKey) {
  if (!hasColorRules) {
    return null;
  }
  const seriesEntry = seriesForColorRule.find(
    (s) => dimensionKey ? s._originalData === categoryData && s._dimensionKey === dimensionKey : s._originalData === categoryData && !s._dimensionKey
  );
  return seriesEntry ? colorRules.for(seriesEntry) : null;
}
function determineColorForDimension(colorRuleResult, originalPaletteColor, dimensionKey, paletteColorPosition) {
  if (isColorToken(colorRuleResult)) {
    return {
      color: colorRuleResult,
      legendColor: originalPaletteColor
    };
  }
  if (isColorPalette(colorRuleResult)) {
    return {
      color: (0, import_get_color_palette.getColorPalette)(
        dimensionKey,
        paletteColorPosition,
        colorRuleResult
      ),
      legendColor: originalPaletteColor
    };
  }
  return {
    color: originalPaletteColor,
    legendColor: void 0
  };
}
function getDimensionColor({
  colorByDimension,
  categoryColor,
  categoryName,
  paletteColorPosition,
  colorPalette,
  colorPaletteMode,
  colorRuleResult
}) {
  const cachedColor = colorByDimension.get(categoryName);
  if (cachedColor) {
    return cachedColor;
  }
  if (categoryColor) {
    return { Default: categoryColor };
  }
  const appliedColorRule = colorRuleResult;
  if (typeof appliedColorRule === "string") {
    return (0, import_get_color_palette.getColorPalette)(
      categoryName,
      paletteColorPosition,
      appliedColorRule,
      colorPaletteMode
    );
  }
  if (!(0, import_lodash_es.isNull)(appliedColorRule) && typeof appliedColorRule === "object") {
    return appliedColorRule;
  }
  if (colorPalette) {
    return (0, import_get_color_palette.getColorPalette)(
      categoryName,
      paletteColorPosition,
      colorPalette,
      colorPaletteMode
    );
  }
  return { Default: import_constants.DEFAULT_SCALE_FALLBACK_COLOR };
}
function handleSingleDimension({
  categoryData,
  rest,
  categoryName,
  categoryColor,
  colorByDimension,
  colorRules,
  colorPalette,
  colorPaletteMode,
  paletteColorPosition,
  uniqueDimensions,
  dimensionMap,
  newCategory,
  seriesForColorRule,
  hasColorRules
}) {
  const value = categoryData.value;
  const dimensionName = categoryName;
  const id = crypto.randomUUID();
  const colorRuleResult = getColorRuleResult(
    hasColorRules,
    colorRules,
    seriesForColorRule,
    categoryData
  );
  const color = getDimensionColor({
    colorByDimension,
    categoryColor,
    categoryName,
    paletteColorPosition,
    colorPalette,
    colorPaletteMode,
    colorRuleResult
  });
  colorByDimension.set(dimensionName, color);
  const newDimension = {
    ...rest,
    id,
    name: dimensionName,
    value
  };
  const dimensionMetadata = {
    id,
    name: dimensionName,
    value,
    color
  };
  uniqueDimensions.add(dimensionName);
  dimensionMap.set(newDimension, dimensionMetadata);
  newCategory.dimensions.push(newDimension);
}
function handleMultipleDimensions({
  categoryData,
  rest,
  categoryName,
  colorByDimension,
  colorByCategory,
  colorRules,
  colorPalette,
  categoricalColorPalette,
  uniqueDimensions,
  dimensionMap,
  newCategory,
  seriesForColorRule,
  hasColorRules
}) {
  Object.entries(categoryData.value).forEach(([dimensionKey, value], index) => {
    const colorRuleCategoryData = getColorRuleResult(
      hasColorRules,
      colorRules,
      seriesForColorRule,
      categoryData,
      dimensionKey
    );
    const paletteColorPosition = categoricalColorPalette ? index : 0;
    const colorByCategoryName = colorByCategory.get(categoryName);
    const dimensionColor = colorByCategoryName?.[dimensionKey];
    let originalPaletteColor;
    if (dimensionColor) {
      originalPaletteColor = dimensionColor;
    } else if (colorPalette && ((0, import_is_valid_color_palette.isValidColorPalette)(colorPalette) || (0, import_color_palette.isCustomColorPalette)(colorPalette))) {
      originalPaletteColor = (0, import_get_color_palette.getColorPalette)(
        dimensionKey,
        paletteColorPosition,
        colorPalette
      );
    } else {
      originalPaletteColor = { Default: import_constants.DEFAULT_SCALE_FALLBACK_COLOR };
    }
    const { color, legendColor } = determineColorForDimension(
      colorRuleCategoryData,
      originalPaletteColor,
      dimensionKey,
      paletteColorPosition
    );
    if (!colorByDimension.has(dimensionKey)) {
      colorByDimension.set(dimensionKey, legendColor ?? color);
    }
    const categoryColorByCategory = colorByCategory.get(categoryName) ?? {};
    categoryColorByCategory[dimensionKey] = legendColor ?? color;
    colorByCategory.set(categoryName, categoryColorByCategory);
    const id = crypto.randomUUID();
    const name = dimensionKey;
    const newDimension = {
      ...rest,
      id,
      name,
      value,
      category: categoryName
    };
    const dimensionMetadata = {
      id,
      name,
      value,
      color,
      legendColor
    };
    uniqueDimensions.add(dimensionKey);
    dimensionMap.set(newDimension, dimensionMetadata);
    newCategory.dimensions.push(newDimension);
  });
}
