// @dynatrace/strato-components/charts/categorical-bar/utils/iterate-slots.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var iterate_slots_exports = {};
__export(iterate_slots_exports, {
  iterateSlots: () => iterateSlots
});
module.exports = __toCommonJS(iterate_slots_exports);
var import_react = require("react");
var import_Intent = require("../../../core/slots/Intent/Intent.js");
var import_is_slot = require("../../../core/utils/is-slot.js");
var import_ColorRule = require("../../core/slots/ColorRule.js");
var import_ChartToolbar = require("../../core/slots/toolbar/ChartToolbar.js");
var import_CompactToolbar = require("../../core/slots/toolbar/CompactToolbar.js");
var import_DownloadData = require("../../core/slots/toolbar/DownloadData.js");
var import_extract_behavior_tracking_props = require("../../core/utils/extract-behavior-tracking-props.js");
var import_CategoryAxis = require("../slots/CategoryAxis.js");
var import_DownloadCSV = require("../slots/DownloadCSV.js");
var import_Legend = require("../slots/Legend.js");
var import_ThresholdIndicator = require("../slots/ThresholdIndicator.js");
var import_Toolbar = require("../slots/Toolbar.js");
var import_Tooltip = require("../slots/Tooltip.js");
var import_ValueAxis = require("../slots/ValueAxis.js");
const iterateChartToolbarSlots = (children) => {
  let slots = {};
  import_react.Children.forEach(children, (childInteractions) => {
    if ((0, import_is_slot.isSlot)(childInteractions, import_DownloadData.DownloadData)) {
      slots = { ...slots, downloadData: { enabled: true } };
    }
  });
  return slots;
};
const iterateSlots = (children) => {
  let slots = {
    legend: {},
    tooltip: {},
    valueAxis: {},
    categoryAxis: {},
    thresholds: [],
    toolbar: {},
    colorRules: [],
    intents: []
  };
  import_react.Children.forEach(children, (child) => {
    if ((0, import_is_slot.isSlot)(child, import_Legend.Legend)) {
      const legend = child.props;
      slots = {
        ...slots,
        legend
      };
    } else if ((0, import_is_slot.isSlot)(child, import_CategoryAxis.CategoryAxis)) {
      const categoryAxis = child.props;
      slots = {
        ...slots,
        categoryAxis
      };
    } else if ((0, import_is_slot.isSlot)(child, import_ValueAxis.ValueAxis)) {
      const valueAxis = child.props;
      slots = {
        ...slots,
        valueAxis
      };
    } else if ((0, import_is_slot.isSlot)(child, import_Tooltip.Tooltip)) {
      const tooltip = child.props;
      slots = {
        ...slots,
        tooltip
      };
    } else if ((0, import_is_slot.isSlot)(child, import_ThresholdIndicator.ThresholdIndicator)) {
      const thresholdProps = child.props;
      slots.thresholds.push(thresholdProps);
    } else if ((0, import_is_slot.isSlot)(child, import_ChartToolbar.ChartToolbar)) {
      const { children: children2, ...toolbar } = child.props;
      const toolbarChildren = iterateChartToolbarSlots(children2);
      slots = {
        ...slots,
        toolbar: {
          ...slots.toolbar,
          ...toolbar,
          ...toolbarChildren
        }
      };
    } else if ((0, import_is_slot.isSlot)(child, import_Intent.Intent)) {
      slots.intents.push(child.props);
    } else if ((0, import_is_slot.isSlot)(child, import_ColorRule.ColorRule)) {
      const colorRule = child.props;
      if ((0, import_ColorRule.isValidColorRule)(colorRule)) {
        slots.colorRules.push(colorRule);
      }
    } else if ((0, import_is_slot.isSlot)(child, import_Toolbar.CategoricalBarChartToolbarSlot)) {
      slots.toolbar = {
        ...slots.toolbar,
        ...(0, import_CompactToolbar.sanitizeProps)(child.props),
        ...(0, import_extract_behavior_tracking_props.extractBehaviorTrackingProps)(child.props)
      };
    } else if ((0, import_is_slot.isSlot)(child, import_DownloadCSV.CategoricalBarChartDownloadCSVSlot)) {
      slots.toolbar.downloadData ||= { enabled: false };
      slots.toolbar.downloadData.enabled = true;
    }
  });
  return slots;
};
