// @dynatrace/strato-components/filters/timeframe-selector/components/custom-timeframe-selector/CustomTimeframeSelector.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CustomTimeframeSelector_exports = {};
__export(CustomTimeframeSelector_exports, {
  CustomTimeframeSelector: () => CustomTimeframeSelector
});
module.exports = __toCommonJS(CustomTimeframeSelector_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_breakpoints_js = __toESM(require("@dynatrace/strato-design-tokens/breakpoints-js"));
var customTimeframeSelectorStyles = __toESM(require("./CustomTimeframeSelector.sty.js"));
var import_Button = require("../../../../buttons/button/Button.js");
var import_useId = require("../../../../core/hooks/useId.js");
var import_useTimeframeStore = require("../../../../core/time/timeframeStore/useTimeframeStore.js");
var import_validation = require("../../../../core/time/validation.js");
var import_DateTimeRangePickerRoot = require("../../../../forms/date-time-range-picker/DateTimeRangePickerRoot.js");
var import_FormFieldMessages = require("../../../../forms/forms-core/components/form-field-messages/FormFieldMessages.js");
var import_Label = require("../../../../forms/forms-core/components/label/Label.js");
var import_FormFieldContext = require("../../../../forms/forms-core/contexts/FormFieldContext.js");
var import_FormFieldMessageContext = require("../../../../forms/forms-core/contexts/FormFieldMessageContext.js");
var import_useFormFieldContext = require("../../../../forms/forms-core/hooks/useFormFieldContext.js");
var import_useFormFieldMessagesContext = require("../../../../forms/forms-core/hooks/useFormFieldMessagesContext.js");
var import_formfield_messages_state_reducer = require("../../../../forms/forms-core/state/formfield-messages-state-reducer.js");
var import_helpers = require("../../../../forms/forms-core/time/helpers.js");
var import_timeParser = require("../../../../forms/forms-core/time/timeParser.js");
var import_Flex = require("../../../../layouts/flex/Flex.js");
var import_useBreakpoint = require("../../../../layouts/hooks/useBreakpoint.js");
var import_InputGroup = require("../../../../layouts/input-group/InputGroup.js");
var import_timeframeSelectorMessages = require("../../constants/timeframeSelectorMessages.js");
var import_presets_helpers = require("../../utils/presets-helpers.js");
var import_timeframe_selection = require("../../utils/timeframe-selection.js");
const CustomTimeframeSelector = () => {
  const ref = (0, import_useTimeframeStore.useTimeframeSelectorStore)(
    (state2) => state2.dateTimeRangePickerRef
  );
  const intl = (0, import_react_intl.useIntl)();
  const value = (0, import_useTimeframeStore.useTimeframeSelectorStore)(
    (state2) => state2.dateTimeRangePickerTimeframe
  );
  const onChange = (0, import_useTimeframeStore.useTimeframeSelectorStore)(
    (state2) => state2.setTimeframeSelectorTimeframe
  );
  const min = (0, import_useTimeframeStore.useTimeframeSelectorStore)((state2) => state2.min);
  const max = (0, import_useTimeframeStore.useTimeframeSelectorStore)((state2) => state2.max);
  const precision = (0, import_useTimeframeStore.useTimeframeSelectorStore)((state2) => state2.precision);
  const required = (0, import_useTimeframeStore.useTimeframeSelectorStore)((state2) => state2.required);
  const dateTimeRangePickerError = (0, import_useTimeframeStore.useTimeframeSelectorStore)(
    (state2) => state2.dateTimeRangePickerError
  );
  const formatString = (0, import_useTimeframeStore.useTimeframeSelectorStore)((state2) => state2.formatString);
  const {
    state: { showError }
  } = (0, import_useFormFieldMessagesContext.useFormFieldMessagesContext)();
  const rootRef = (0, import_react.useRef)(null);
  const dateTimeRangePickerId = (0, import_useId.useId)();
  const hintId = (0, import_useId.useId)();
  const labelId = (0, import_useId.useId)();
  const { from, to } = (0, import_timeParser.parseValue)(value);
  const { required: contextRequired } = (0, import_useFormFieldContext.useFormFieldContext)();
  const callback = (0, import_react.useCallback)(
    (state2, action) => (0, import_formfield_messages_state_reducer.formFieldMessagesReducer)(state2, action),
    []
  );
  const [state, dispatch] = (0, import_react.useReducer)(callback, {
    customMessages: [],
    formControlMessages: [],
    showError,
    isTouched: showError
  });
  const applyButtonDataDtName = "Apply";
  const { valueFrom, valueTo } = (0, import_timeParser.getFromToValues)(value);
  const applyButtonDataDtProperties = `valueFrom:${valueFrom},valueTo:${valueTo}`;
  const previousErrorRef = (0, import_react.useRef)();
  (0, import_react.useEffect)(() => {
    previousErrorRef.current = (0, import_validation.validate)(value, min, max, {
      formMessageDispatch: dispatch,
      formerError: previousErrorRef.current,
      precision,
      showError,
      checkOrder: true,
      ref,
      expectedFormat: formatString,
      valueMode: "dateTimePicker",
      required: required || contextRequired,
      errorMessages: import_timeframeSelectorMessages.TIMEFRAMESELECTOR_ERROR_MESSAGES
    });
    if (!required && !contextRequired) {
      if (!from && to?.value === "now()") {
        dispatch({
          type: import_formfield_messages_state_reducer.FormFieldMessagesAction.ADDMESSAGE,
          payload: {
            id: hintId,
            variant: "error",
            message: intl.formatMessage(
              import_timeframeSelectorMessages.TIMEFRAMESELECTOR_ERROR_MESSAGES.missingFrom
            ),
            messageType: "formcontrol",
            validityStateKey: "customError"
          }
        });
        return;
      }
      if (!to && from?.value === "now()") {
        dispatch({
          type: import_formfield_messages_state_reducer.FormFieldMessagesAction.ADDMESSAGE,
          payload: {
            id: hintId,
            variant: "error",
            message: intl.formatMessage(
              import_timeframeSelectorMessages.TIMEFRAMESELECTOR_ERROR_MESSAGES.missingTo
            ),
            messageType: "formcontrol",
            validityStateKey: "customError"
          }
        });
        return;
      }
    }
  }, [
    value,
    min,
    max,
    dispatch,
    precision,
    formatString,
    required,
    contextRequired,
    showError,
    ref
  ]);
  const handleApply = (0, import_react.useCallback)(() => {
    dispatch({
      type: import_formfield_messages_state_reducer.FormFieldMessagesAction.SETISTOUCHED,
      payload: { isTouched: true }
    });
    if (dateTimeRangePickerError) {
      return;
    }
    const currentValue = ref.current?.getCurrentValue();
    const refFrom = currentValue?.from?.value;
    const refTo = currentValue?.to?.value;
    let currentTimeframe = value;
    const onlyFromIsEmpty = !refFrom && !!refTo && refTo !== "now()";
    const onlyToIsEmpty = !refTo && !!refFrom && refFrom !== "now()";
    if (onlyFromIsEmpty || onlyToIsEmpty) {
      const from3 = ref.current?.inputRef.from?.value || "now()";
      const to3 = ref.current?.inputRef.to?.value || "now()";
      const parsedTimeframe = (0, import_timeParser.parseValue)({ from: from3, to: to3 }, false, precision);
      if (parsedTimeframe.from && parsedTimeframe.to) {
        if ((0, import_timeframe_selection.isOutsideRange)(
          {
            from: parsedTimeframe.from.value,
            to: parsedTimeframe.to.value
          },
          min?.value,
          max?.value
        )) {
          return;
        }
        const sortedTimeframe = (0, import_helpers.sortTimeframeValues)({
          from: parsedTimeframe.from,
          to: parsedTimeframe.to
        });
        if (sortedTimeframe.from && sortedTimeframe.to) {
          currentTimeframe = {
            from: sortedTimeframe.from,
            to: sortedTimeframe.to
          };
        }
      }
    }
    const { from: from2, to: to2 } = currentTimeframe ? (0, import_timeParser.parseValue)(currentTimeframe, false, precision) : { from: null, to: null };
    if (from2 && to2) {
      const timeframe = { from: from2, to: to2 };
      onChange(timeframe);
      (0, import_presets_helpers.saveRecentTimeframe)(timeframe);
      return;
    }
    if (!required && !contextRequired) {
      const hasFromValue = Boolean(refFrom);
      const hasToValue = Boolean(refTo);
      if (!hasFromValue && !hasToValue) {
        onChange(null);
      }
    }
  }, [
    dateTimeRangePickerError,
    ref,
    value,
    precision,
    required,
    contextRequired,
    min,
    max,
    onChange
  ]);
  const handleKeyDown = (0, import_react.useCallback)(
    (e) => {
      if (e.key !== "Enter" || // Checks if the event target is a child of the DateTimeRangePicker root element ref.
      !ref.current?.element?.contains(e.target)) {
        return;
      }
      ref.current?.validate();
      handleApply();
    },
    [handleApply, ref]
  );
  const isScreenLargerThanMobile = (0, import_useBreakpoint.useBreakpoint)(
    `(min-width: ${import_breakpoints_js.default.Max.Mobile}px)`,
    true
  );
  (0, import_react.useEffect)(() => {
    const timeBoundaryMessageData = (0, import_validation.getTimeBoundaryMessageData)(
      {
        from,
        to
      },
      min,
      max,
      precision,
      false
    );
    if (!timeBoundaryMessageData) {
      if (!required && !contextRequired) {
        if (!from && to?.value === "now()") {
          dispatch({
            type: import_formfield_messages_state_reducer.FormFieldMessagesAction.ADDMESSAGE,
            payload: {
              id: hintId,
              variant: "error",
              message: intl.formatMessage(
                import_timeframeSelectorMessages.TIMEFRAMESELECTOR_ERROR_MESSAGES.missingFrom
              ),
              messageType: "formcontrol",
              validityStateKey: "customError"
            }
          });
          return;
        }
        if (!to && from?.value === "now()") {
          dispatch({
            type: import_formfield_messages_state_reducer.FormFieldMessagesAction.ADDMESSAGE,
            payload: {
              id: hintId,
              variant: "error",
              message: intl.formatMessage(
                import_timeframeSelectorMessages.TIMEFRAMESELECTOR_ERROR_MESSAGES.missingTo
              ),
              messageType: "formcontrol",
              validityStateKey: "customError"
            }
          });
          return;
        }
      }
      dispatch({
        type: import_formfield_messages_state_reducer.FormFieldMessagesAction.REMOVEMESSAGE,
        payload: {
          id: hintId,
          messageType: "formcontrol"
        }
      });
      return;
    }
    dispatch({
      type: import_formfield_messages_state_reducer.FormFieldMessagesAction.ADDMESSAGE,
      payload: {
        id: hintId,
        variant: "hint",
        message: intl.formatMessage(
          import_timeframeSelectorMessages.TIMEFRAMESELECTOR_ERROR_MESSAGES[timeBoundaryMessageData.errorKey],
          timeBoundaryMessageData.values
        ),
        messageType: "formcontrol",
        validityStateKey: "customError"
      }
    });
  }, [
    min,
    max,
    dispatch,
    intl,
    precision,
    hintId,
    from,
    to,
    required,
    contextRequired
  ]);
  const messageContextValue = (0, import_react.useMemo)(
    () => ({
      dispatch,
      state
    }),
    [dispatch, state]
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { ref: rootRef, className: customTimeframeSelectorStyles.overlayRoot, children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_FormFieldMessageContext.FormFieldMessageContext.Provider, { value: messageContextValue, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { className: customTimeframeSelectorStyles.wrapper, children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { onKeyDown: handleKeyDown, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_FormFieldContext._FormFieldContext.Provider, { value: {}, children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
        "fieldset",
        {
          className: customTimeframeSelectorStyles.fieldset,
          "aria-labelledby": labelId,
          children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_Label.Label,
              {
                htmlFor: dateTimeRangePickerId,
                required: required || contextRequired,
                id: labelId,
                children: intl.formatMessage(
                  import_timeframeSelectorMessages.TIMEFRAME_SELECTOR_MESSAGES.timeframeLabel
                )
              }
            ),
            /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_InputGroup.InputGroup, { style: { gridAutoColumns: "1fr auto" }, children: [
              /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_DateTimeRangePickerRoot.DateTimeRangePickerRoot,
                {
                  ref,
                  id: dateTimeRangePickerId,
                  required: required || contextRequired,
                  style: { gridTemplateColumns: "1fr auto" },
                  onBlur: (event) => {
                    const hasPresetItemSelected = event.relatedTarget?.getAttribute("role") === "option";
                    const isActiveElementInside = ref.current?.element?.contains(document.activeElement);
                    if (hasPresetItemSelected || isActiveElementInside) {
                      return;
                    }
                    dispatch({
                      type: import_formfield_messages_state_reducer.FormFieldMessagesAction.SETISTOUCHED,
                      payload: { isTouched: true }
                    });
                  }
                }
              ),
              isScreenLargerThanMobile && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                import_Button.Button,
                {
                  color: "primary",
                  variant: "accent",
                  type: "submit",
                  onClick: handleApply,
                  style: {
                    alignSelf: "end",
                    borderTopLeftRadius: 0,
                    borderBottomLeftRadius: 0
                  },
                  "data-dt-name": applyButtonDataDtName,
                  "data-dt-properties": applyButtonDataDtProperties,
                  children: intl.formatMessage(
                    import_timeframeSelectorMessages.TIMEFRAME_SELECTOR_MESSAGES.applyButton
                  )
                }
              )
            ] })
          ]
        }
      ) }) }),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_FormFieldMessages.FormFieldMessages, {})
    ] }),
    !isScreenLargerThanMobile && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Button.Button,
      {
        color: "primary",
        variant: "accent",
        type: "submit",
        style: { alignSelf: "end" },
        onClick: handleApply,
        "data-dt-name": applyButtonDataDtName,
        "data-dt-properties": applyButtonDataDtProperties,
        children: intl.formatMessage(import_timeframeSelectorMessages.TIMEFRAME_SELECTOR_MESSAGES.applyButton)
      }
    )
  ] }) });
};
CustomTimeframeSelector["displayName"] = "CustomTimeframeSelector";
