// @dynatrace/strato-components/charts/timeseries/hooks/store/selectors.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var selectors_exports = {};
__export(selectors_exports, {
  useAnnotationsState: () => useAnnotationsState,
  useCurrentDomainState: () => useCurrentDomainState,
  useExploreState: () => useExploreState,
  useFilteredSeriesIdsState: () => useFilteredSeriesIdsState,
  useHighlightedSeriesIdState: () => useHighlightedSeriesIdState,
  useInteractionModeState: () => useInteractionModeState,
  useSelectState: () => useSelectState,
  useSelector: () => useSelector,
  useSetState: () => useSetState,
  useThresholdsState: () => useThresholdsState,
  useZoomState: () => useZoomState
});
module.exports = __toCommonJS(selectors_exports);
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_selector = require("../../../../core/store/selector.js");
var import_TimeseriesChartState_context = require("../../context/TimeseriesChartState.context.js");
const useSelector = (selector, compareFn) => (0, import_selector.useStoreSelector)(import_TimeseriesChartState_context.TimeseriesChartStateContext, selector, compareFn);
const useZoomState = () => useSelector((state) => state.zoom);
const useAnnotationsState = () => useSelector((state) => state.annotations);
const useThresholdsState = () => useSelector((state) => state.thresholds);
const useFilteredSeriesIdsState = () => useSelector((state) => state.filteredSeriesIds);
const useHighlightedSeriesIdState = () => useSelector((state) => state.highlightedSeriesId);
const useCurrentDomainState = () => useSelector((state) => state.currentDomain);
const useInteractionModeState = () => useSelector((state) => state.interactionMode);
const useExploreState = () => useSelector((state) => state.explore, import_lodash_es.isEqual);
const useSelectState = () => useSelector((state) => state.select);
const useSetState = () => {
  const { setState } = (0, import_react.useContext)(import_TimeseriesChartState_context.TimeseriesChartStateContext);
  return setState;
};
