// @dynatrace/strato-components/tables/DataTable/DataTable.sty.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DataTable_css_exports = {};
__export(DataTable_css_exports, {
  dataTable: () => dataTable,
  firstBodyRow: () => firstBodyRow,
  fullWidth: () => fullWidth,
  lastBodyRow: () => lastBodyRow,
  notResizing: () => notResizing,
  pinnedCellInDragOverlay: () => pinnedCellInDragOverlay,
  prefixColumn: () => prefixColumn
});
module.exports = __toCommonJS(DataTable_css_exports);
var import_DataTable_css_ts_vanilla = require("./DataTable.css");
var dataTable = "_14sp6az0-3-6-0";
var firstBodyRow = "_14sp6az5-3-6-0";
var fullWidth = "_14sp6az1-3-6-0";
var lastBodyRow = "_14sp6az6-3-6-0";
var notResizing = "_14sp6az2-3-6-0";
var pinnedCellInDragOverlay = "_14sp6az4-3-6-0";
var prefixColumn = "_14sp6az3-3-6-0";
