// @dynatrace/strato-components/tables/DataTable/DataTable.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DataTable_exports = {};
__export(DataTable_exports, {
  DataTable: () => DataTable
});
module.exports = __toCommonJS(DataTable_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_DataTableDefaultCell = require("./components/DataTableDefaultCell.js");
var import_DataTableEmptyState = require("./components/DataTableEmptyState.js");
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_Grid = require("../../layouts/grid/Grid.js");
var import_useDataTableConfig = require("../hooks/useDataTableConfig.js");
var import_DataTableVirtualizationContainer = require("./components/DataTableVirtualizationContainer.js");
var import_DataTableSelectedRowsActions = require("./components/TableActions/DataTableSelectedRowsActions.js");
var import_DataTableSelectionChip = require("./components/TableActions/DataTableSelectionChip.js");
var import_DataTableTableActions = require("./components/TableActions/DataTableTableActions.js");
var import_DataTableTableActionsWrapper = require("./components/TableActions/DataTableTableActionsWrapper.js");
var import_DataTableColumnOrderSettings = require("./components/Toolbar/ColumnSettings/DataTableColumnOrderSettings.js");
var import_DataTableColumnPinningSettings = require("./components/Toolbar/ColumnSettings/DataTableColumnPinningSettings.js");
var import_DataTableColumnSettingsModal = require("./components/Toolbar/ColumnSettings/DataTableColumnSettingsModal.js");
var import_DataTableColumnSettingsTrigger = require("./components/Toolbar/ColumnSettings/DataTableColumnSettingsTrigger.js");
var import_DataTableColumnVisibilitySettings = require("./components/Toolbar/ColumnSettings/DataTableColumnVisibilitySettings.js");
var import_DataTableToolbar = require("./components/Toolbar/DataTableToolbar.js");
var import_DataTableContext = require("./contexts/DataTableContext.js");
var import_core_row_model = require("./core-row-model.js");
var styles = __toESM(require("./DataTable.sty.js"));
var import_ColumnFontStyle = require("./features/ColumnFontStyle/ColumnFontStyle.js");
var import_ColumnOrder = require("./features/ColumnOrder/ColumnOrder.js");
var import_ColumnPinning = require("./features/ColumnPinning/ColumnPinning.js");
var import_ColumnSizing = require("./features/ColumnSizing/ColumnSizing.js");
var import_ColumnVisibility = require("./features/ColumnVisibility/ColumnVisibility.js");
var import_Config = require("./features/Config.js");
var import_DevTools = require("./features/DevTools.js");
var import_Download = require("./features/Download/Download.js");
var import_GanttChart = require("./features/GanttChart/GanttChart.js");
var import_Layout = require("./features/Layout/Layout.js");
var import_LineWrap = require("./features/LineWrap/LineWrap.js");
var import_LoadingState = require("./features/LoadingState/LoadingState.js");
var import_MeterBarChart = require("./features/MeterBarChart/MeterBarChart.js");
var import_Pagination = require("./features/Pagination/Pagination.js");
var import_PaginationInternal = require("./features/Pagination/PaginationInternal.js");
var import_RowDetails = require("./features/RowDetails/RowDetails.js");
var import_RowIdentification = require("./features/RowIdentification/RowIdentification.js");
var import_RowInteractivity = require("./features/RowInteractivity/RowInteractivity.js");
var import_RowOrder = require("./features/RowOrder/RowOrder.js");
var import_RowSelection = require("./features/RowSelection/RowSelection.js");
var import_sorted_row_model = require("./features/Sorting/sorted-row-model.js");
var import_Sorting = require("./features/Sorting/Sorting.js");
var import_SubRows = require("./features/SubRows/SubRows.js");
var thresholdStyles = __toESM(require("./features/Thresholds/Thresholds.sty.js"));
var import_Thresholds = require("./features/Thresholds/Thresholds.js");
var import_CellActions = require("./features/UserActions/CellActions.js");
var import_ColumnActions = require("./features/UserActions/ColumnActions.js");
var import_RowActions = require("./features/UserActions/RowActions.js");
var import_useUserActions = require("./features/UserActions/useUserActions.js");
var import_VisualVariants = require("./features/VisualVariants/VisualVariants.js");
var import_useBuiltInColumns = require("./hooks/useBuiltInColumns.js");
var import_useColumnRemeasure = require("./hooks/useColumnRemeasure.js");
var import_useColumns = require("./hooks/useColumns.js");
var import_useDataTableSlots = require("./hooks/useDataTableSlots.js");
var import_usePaginationEmptyPageCorrection = require("./hooks/usePaginationEmptyPageCorrection.js");
var import_usePaginationSlot = require("./hooks/usePaginationSlot.js");
var import_useTable = require("./hooks/useTable/useTable.js");
var import_useTableRootResize = require("./hooks/useTableRootResize.js");
var import_useToolbarSlot = require("./hooks/useToolbarSlot.js");
var import_src = require("./tanstack-table-core-override/src/index.js");
var import_Intent = require("../../core/slots/Intent/Intent.js");
var import_extract_data_attributes = require("../utils/extract-data-attributes.js");
const COMPONENT_NAME = "DataTable";
function InternalDataTable(props, forwardedRef) {
  const {
    children,
    data: dataProps,
    columns: columnsProps,
    style: consumerStyle,
    className: consumerClassName,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const ariaLabelingProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(restProps);
  const dataAttributes = (0, import_extract_data_attributes.extractDataAttributes)(restProps);
  const slots = (0, import_useDataTableSlots.useDataTableSlots)(children);
  const {
    toolbarSlots,
    toolbarContent,
    toolbarItemProps,
    toolbarRemainingProps,
    openColumnSettings,
    columnSettingsModalRef
  } = (0, import_useToolbarSlot.useToolbarSlot)(slots.toolbarProps);
  const {
    paginationConsumerId,
    paginationDataTestId,
    paginationConsumerStyle,
    paginationConsumerClassName,
    paginationDataAttributes
  } = (0, import_usePaginationSlot.usePaginationSlot)(slots.pagination);
  const [columns, consumerColumnDefMap] = (0, import_useColumns.useColumns)(
    columnsProps
  );
  const configFullWidthOption = (0, import_useDataTableConfig.useDataTableConfigElement)("fullWidth");
  const fullWidth = props.fullWidth ?? configFullWidthOption;
  const configFullHeightOption = (0, import_useDataTableConfig.useDataTableConfigElement)("fullHeight");
  const fullHeight = props.fullHeight ?? configFullHeightOption;
  const options = {
    columns,
    data: dataProps,
    // In order to keep the types in our own realm, the cast to align with our
    // CoreOptions is necessary.
    getCoreRowModel: (0, import_core_row_model.getCoreRowModel)(),
    getSortedRowModel: (0, import_sorted_row_model.getSortedRowModel)(),
    // In order to keep the types in our own realm, the cast to align with our
    // CoreOptions is necessary.
    getExpandedRowModel: (0, import_src.getExpandedRowModel)()
  };
  (0, import_RowIdentification.useRowIdentification)(props, options);
  (0, import_Layout.useLayout)(props, options);
  (0, import_LineWrap.useColumnLineWrap)(props, options);
  (0, import_ColumnFontStyle.useColumnFontStyle)(props, options);
  (0, import_Thresholds.useThresholds)(props, options);
  (0, import_VisualVariants.useVisualVariant)(props, options);
  (0, import_useUserActions.useUserActions)(props, options, consumerColumnDefMap);
  (0, import_Pagination.usePagination)(props, options);
  (0, import_LoadingState.useLoadingState)(props, options);
  (0, import_ColumnSizing.useColumnSizing)(props, options);
  (0, import_Sorting.useSorting)(props, options);
  (0, import_ColumnVisibility.useColumnVisibility)(props, options);
  (0, import_ColumnOrder.useColumnOrder)(props, options);
  (0, import_RowSelection.useRowSelection)(props, options);
  (0, import_SubRows.useSubRows)(props, options);
  (0, import_RowDetails.useRowDetails)(props, options);
  (0, import_RowInteractivity.useRowInteractivity)(props, options);
  (0, import_GanttChart.useGanttChart)(props, options);
  (0, import_MeterBarChart.useMeterBarChartMaxValuesCalc)(props, options);
  (0, import_RowOrder.useRowOrder)(props, options);
  (0, import_ColumnPinning.useColumnPinning)(props, options);
  (0, import_Download.useDownloadData)(
    // eslint-disable-next-line @typescript-eslint/no-deprecated
    { onDownloadData: toolbarItemProps.onDownloadData ?? props.onDownloadData },
    options
  );
  options.columns = (0, import_useBuiltInColumns.useBuiltInColumns)(options);
  const table = (0, import_useTable.useTable)(options);
  const tableRef = (0, import_Config.useDataTableConfigRef)(
    forwardedRef,
    columnSettingsModalRef,
    table,
    props
  );
  const rootRef = (0, import_useTableRootResize.useTableRootResize)(table.remeasureRows, table.setState);
  const mergedRefs = (0, import_useMergeRefs.useMergeRefs)([rootRef, tableRef]);
  (0, import_useColumnRemeasure.useColumnRemeasure)(table);
  (0, import_usePaginationEmptyPageCorrection.usePaginationEmptyPageCorrection)(table);
  const DevTools = (0, import_DevTools._getDevToolsSlot)(props, table);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "div",
      {
        ref: mergedRefs,
        className: (0, import_clsx.default)(
          styles.dataTable,
          {
            [styles.fullWidth]: fullWidth,
            [styles.notResizing]: table.getState().columnSizingInfo.isResizingColumn === false
          },
          {
            [thresholdStyles.hasPillThresholds]: table.options.hasThresholds && table.options.rowThresholds?.some(
              (threshold) => threshold.type === "pill"
            )
          },
          consumerClassName
        ),
        style: consumerStyle,
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        ...ariaLabelingProps,
        ...behavioralTrackingProps,
        ...dataAttributes,
        "data-dt-component": COMPONENT_NAME,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_DataTableContext.DataTableContextProvider, { value: table, children: [
          (toolbarContent || slots.tableActions || slots.selectedRowsActions) && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
            import_Grid.Grid,
            {
              gridTemplateColumns: "1fr auto",
              paddingBottom: 4,
              gap: toolbarContent ? 8 : 0,
              children: [
                /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  import_DataTableTableActionsWrapper.DataTableTableActionsWrapper,
                  {
                    tableActions: slots.tableActions,
                    selectedRowsActions: slots.selectedRowsActions
                  }
                ),
                /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  import_DataTableContext.DataTableColumnSettingsContextProvider,
                  {
                    value: { openColumnSettings },
                    children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_DataTableToolbar.DataTableToolbar, { ...toolbarRemainingProps, children: toolbarContent })
                  }
                )
              ]
            }
          ),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_DataTableColumnSettingsModal.DataTableColumnSettingsModal,
            {
              ref: columnSettingsModalRef,
              visibilitySettings: toolbarSlots.visibilitySettings,
              columnOrderSettings: toolbarSlots.columnOrderSettings,
              columnPinningSettings: toolbarSlots.columnPinningSettings,
              resetColumnVisibility: toolbarItemProps.resetColumnVisibility,
              resetColumnOrder: toolbarItemProps.resetColumnOrder,
              resetColumnPinning: toolbarItemProps.resetColumnPinning
            }
          ),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_DataTableVirtualizationContainer.DataTableVirtualizationContainer,
            {
              table,
              emptyState: slots.emptyState,
              fullHeight: fullHeight ?? false
            }
          ),
          table.options.enablePagination === true ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_PaginationInternal.DataTablePaginationInternal,
            {
              id: paginationConsumerId,
              "data-testid": paginationDataTestId,
              className: paginationConsumerClassName,
              style: paginationConsumerStyle,
              ...paginationDataAttributes
            }
          ) : null
        ] })
      }
    ),
    DevTools
  ] });
}
InternalDataTable.displayName = COMPONENT_NAME;
const DataTableFwd = /* @__PURE__ */ (0, import_react.forwardRef)(InternalDataTable);
const DataTable = Object.assign(DataTableFwd, {
  Pagination: import_Pagination.DataTablePagination,
  ColumnActions: import_ColumnActions.DataTableColumnActions,
  CellActions: import_CellActions.DataTableCellActions,
  RowActions: import_RowActions.DataTableRowActions,
  SelectedRowsActions: import_DataTableSelectedRowsActions.DataTableSelectedRowsActions,
  DefaultCell: import_DataTableDefaultCell.DataTableDefaultCell,
  EmptyState: import_DataTableEmptyState.DataTableEmptyState,
  TableActions: import_DataTableTableActions.DataTableTableActions,
  DevTools: import_DevTools.DataTableDevTools,
  Toolbar: import_DataTableToolbar.DataTableToolbar,
  VisibilitySettings: import_DataTableColumnVisibilitySettings.DataTableColumnVisibilitySettings,
  ColumnOrderSettings: import_DataTableColumnOrderSettings.DataTableColumnOrderSettings,
  ColumnPinningSettings: import_DataTableColumnPinningSettings.DataTableColumnPinningSettings,
  ColumnSettingsTrigger: import_DataTableColumnSettingsTrigger.DataTableColumnSettingsTrigger,
  SelectionChip: import_DataTableSelectionChip.DataTableSelectionChip,
  DownloadData: import_Download.DataTableDownload,
  Intent: import_Intent.Intent,
  ExpandableRow: import_RowDetails.DataTableExpandableRowTemplate,
  ExpandableRowWrapper: import_RowDetails.DataTableExpandableRowWrapper,
  LineWrap: import_LineWrap.DataTableGlobalLineWrapAction
});
