// @dynatrace/strato-components/forms/number-input-v2/utils/calculations.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var calculations_exports = {};
__export(calculations_exports, {
  applyStep: () => applyStep
});
module.exports = __toCommonJS(calculations_exports);
var import_format = require("./format.js");
var import_validation = require("./validation.js");
function resolveInputValue(value) {
  if (value === null) {
    return 0;
  }
  if (typeof value === "string") {
    if (value.trim() === "") {
      return 0;
    }
    const parsed = (0, import_format.parseNumber)(value);
    return parsed ?? 0;
  }
  return value;
}
function isValidStep(step) {
  return step !== void 0 && !Number.isNaN(step) && step > 0;
}
function computeStepAlignedValue(parsedValue, direction, cleanedStep, baseValue) {
  const remainder = Math.abs((parsedValue - baseValue) % cleanedStep);
  const softEpsilon = Number.EPSILON * 100;
  if (remainder <= softEpsilon || remainder >= cleanedStep - softEpsilon) {
    return void 0;
  }
  const nextValue = parsedValue + direction * cleanedStep;
  if (parsedValue < baseValue) {
    return nextValue + remainder - (direction > 0 ? cleanedStep : 0);
  }
  return nextValue - remainder + (direction < 0 ? cleanedStep : 0);
}
function applyMinConstraint(value, min) {
  if (min === void 0 || Number.isNaN(min)) {
    return value;
  }
  return Math.max(value, min);
}
function applyMaxConstraint(value, max, baseValue, cleanedStep, validStep) {
  if (max === void 0 || Number.isNaN(max)) {
    return value;
  }
  if (validStep && value >= max) {
    return baseValue + cleanedStep * Math.floor((max - baseValue) / cleanedStep);
  }
  return Math.min(value, max);
}
function roundToPrecision(value) {
  const precision = Math.pow(10, 10);
  return Math.round(value * precision) / precision;
}
function applyStep(value, step, stepMultiplier, min, max, direction) {
  const parsedValue = resolveInputValue(value);
  if (Number.isNaN(parsedValue)) {
    return value;
  }
  const validStep = isValidStep(step);
  const cleanedStep = (0, import_validation.sanitizePositiveNumber)(step);
  const effectiveStepSize = Math.trunc((0, import_validation.sanitizePositiveNumber)(stepMultiplier)) * cleanedStep;
  const baseValue = min || 0;
  let nextValue = parsedValue + direction * effectiveStepSize;
  if (validStep) {
    const aligned = computeStepAlignedValue(
      parsedValue,
      direction,
      cleanedStep,
      baseValue
    );
    if (aligned !== void 0) {
      nextValue = aligned;
    }
  }
  nextValue = applyMinConstraint(nextValue, min);
  nextValue = applyMaxConstraint(
    nextValue,
    max,
    baseValue,
    cleanedStep,
    validStep
  );
  return roundToPrecision(nextValue);
}
