// @dynatrace/strato-components/forms/number-input-v2/utils/format.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var format_exports = {};
__export(format_exports, {
  checkFormat: () => checkFormat,
  getLocaleSeparators: () => getLocaleSeparators,
  parseNumber: () => parseNumber
});
module.exports = __toCommonJS(format_exports);
var import_user_preferences = require("@dynatrace-sdk/user-preferences");
function parseNumber(value) {
  if (typeof value === "number") {
    return value;
  }
  const regionalFormat = (0, import_user_preferences.getRegionalFormat)();
  const { decimal: localeDecimalSeparator } = getLocaleSeparators(regionalFormat);
  const normalizedValue = checkFormat(value);
  const decimalNormalizedValue = normalizedValue.replaceAll(
    localeDecimalSeparator,
    "."
  );
  const groupingNormalizedValue = decimalNormalizedValue.replace(
    // Match whitespace-like grouping separators (space, NBSP, narrow NBSP, apostrophe), e.g. "1 234", "1\u00A0234", "1\u202F234", "1'234".
    /[\s\u00A0\u202F']/g,
    ""
  );
  return parseFloat(groupingNormalizedValue);
}
function sanitizeNumberLikeInput(inputValue) {
  let sanitizedValue = inputValue.trim();
  if (/^\(.*\)$/.exec(sanitizedValue)) {
    sanitizedValue = sanitizedValue.slice(1, -1).trim();
  }
  const containsSignCharacter = /[-+]/.exec(sanitizedValue);
  if (containsSignCharacter) {
    sanitizedValue = sanitizedValue.replace(/[-+]/g, "").trim();
  }
  return sanitizedValue.replace(/[^\d.,'\s\u00A0\u202F]/g, "");
}
function countDigitsAroundSeparator(inputValue, separatorIndex) {
  const leftPart = inputValue.slice(0, separatorIndex);
  const rightPart = inputValue.slice(separatorIndex + 1);
  const leftDigitCount = (leftPart.match(/\d/g) || []).length;
  const rightDigitCount = (rightPart.match(/\d/g) || []).length;
  return { leftPart, leftDigitCount, rightDigitCount };
}
function hasLeftGroupingPattern(leftPart, separatorCharacter) {
  const cleanedLeftPart = leftPart.replace(/[\s\u00A0\u202F']/g, "");
  const groupedParts = cleanedLeftPart.split(separatorCharacter);
  if (groupedParts.length <= 1) {
    return false;
  }
  if (!/^\d{1,3}$/.exec(groupedParts[0])) {
    return false;
  }
  for (let partIndex = 1; partIndex < groupedParts.length; partIndex++) {
    if (!/^\d{3}$/.exec(groupedParts[partIndex])) {
      return false;
    }
  }
  return true;
}
function detectWhitespaceGroupingSeparator(inputValue) {
  if (!/[ \u00A0\u202F']/.exec(inputValue)) {
    return null;
  }
  if (inputValue.includes("\u202F")) {
    return "\u202F";
  }
  if (inputValue.includes("\xA0")) {
    return "\xA0";
  }
  if (inputValue.includes("'")) {
    return "'";
  }
  return " ";
}
function inferDotAndCommaSeparators(sanitizedInputValue, fallbackDecimalSeparator) {
  const lastCommaIndex = sanitizedInputValue.lastIndexOf(",");
  const lastDotIndex = sanitizedInputValue.lastIndexOf(".");
  let decimalSeparator = fallbackDecimalSeparator;
  let groupSeparator = null;
  if (lastCommaIndex !== -1 && lastDotIndex !== -1) {
    decimalSeparator = lastCommaIndex > lastDotIndex ? "," : ".";
    groupSeparator = decimalSeparator === "," ? "." : ",";
    return { decimal: decimalSeparator, group: groupSeparator };
  }
  if (lastCommaIndex === -1 && lastDotIndex === -1) {
    return { decimal: decimalSeparator, group: groupSeparator };
  }
  const separatorCharacter = lastCommaIndex !== -1 ? "," : ".";
  const separatorIndex = lastCommaIndex !== -1 ? lastCommaIndex : lastDotIndex;
  const { leftPart, rightDigitCount } = countDigitsAroundSeparator(
    sanitizedInputValue,
    separatorIndex
  );
  const leftPartUsesGrouping = hasLeftGroupingPattern(
    leftPart,
    separatorCharacter
  );
  const hasExactlyThreeDigitsOnRight = rightDigitCount === 3;
  if (hasExactlyThreeDigitsOnRight && (leftPartUsesGrouping || separatorCharacter !== fallbackDecimalSeparator)) {
    groupSeparator = separatorCharacter;
  } else {
    decimalSeparator = separatorCharacter;
  }
  return { decimal: decimalSeparator, group: groupSeparator };
}
function inferSeparators(inputValue, fallbackDecimalSeparator) {
  if (inputValue === null) {
    return { decimal: fallbackDecimalSeparator, group: null };
  }
  const sanitizedInputValue = sanitizeNumberLikeInput(inputValue);
  const inferredDotAndCommaSeparators = inferDotAndCommaSeparators(
    sanitizedInputValue,
    fallbackDecimalSeparator
  );
  const whitespaceGroupingSeparator = detectWhitespaceGroupingSeparator(sanitizedInputValue);
  if (whitespaceGroupingSeparator) {
    return {
      decimal: inferredDotAndCommaSeparators.decimal,
      group: whitespaceGroupingSeparator
    };
  }
  if (!inferredDotAndCommaSeparators.group && (sanitizedInputValue.includes(",") || sanitizedInputValue.includes("."))) {
    const alternateSeparator = inferredDotAndCommaSeparators.decimal === "," ? "." : ",";
    if (sanitizedInputValue.includes(alternateSeparator)) {
      return {
        decimal: inferredDotAndCommaSeparators.decimal,
        group: alternateSeparator
      };
    }
  }
  return inferredDotAndCommaSeparators;
}
function getLocaleSeparators(locale, options = {}) {
  const numberFormatter = new Intl.NumberFormat(locale, {
    useGrouping: true,
    ...options
  });
  const formattedParts = numberFormatter.formatToParts(1234.5);
  let decimalSeparator = ".";
  let groupSeparator = null;
  for (const part of formattedParts) {
    if (part.type === "decimal") {
      decimalSeparator = part.value;
    }
    if (part.type === "group") {
      groupSeparator = part.value;
    }
  }
  return { decimal: decimalSeparator, group: groupSeparator };
}
function removeAllButLast(sourceText, tokenToKeepLast) {
  const lastTokenIndex = sourceText.lastIndexOf(tokenToKeepLast);
  if (lastTokenIndex === -1) {
    return sourceText;
  }
  const prefix = sourceText.slice(0, lastTokenIndex);
  const suffix = sourceText.slice(lastTokenIndex);
  return prefix.replaceAll(tokenToKeepLast, "") + suffix;
}
function checkFormat(value) {
  const regionalFormat = (0, import_user_preferences.getRegionalFormat)();
  const { decimal } = getLocaleSeparators(regionalFormat);
  const { decimal: inferredDecimal, group: inferredGroup } = inferSeparators(
    value,
    decimal
  );
  let normalizedValue = value;
  if (inferredGroup) {
    normalizedValue = normalizedValue.replaceAll(inferredGroup, "");
  }
  if (inferredDecimal && decimal) {
    normalizedValue = normalizedValue.replaceAll(
      inferredDecimal ?? "",
      decimal
    );
    normalizedValue = removeAllButLast(normalizedValue, decimal);
  }
  return normalizedValue;
}
