// @dynatrace/strato-components/forms/number-input-v2/utils/validation.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var validation_exports = {};
__export(validation_exports, {
  addError: () => addError,
  getErrorKey: () => getErrorKey,
  isStepAligned: () => isStepAligned,
  sanitizePositiveNumber: () => sanitizePositiveNumber
});
module.exports = __toCommonJS(validation_exports);
var import_format = require("./format.js");
var import_useFormFieldValidation = require("../../forms-core/hooks/useFormFieldValidation.js");
var import_formfield_messages_state_reducer = require("../../forms-core/state/formfield-messages-state-reducer.js");
function sanitizePositiveNumber(value) {
  return value !== void 0 && !Number.isNaN(value) && value > 0 ? value : 1;
}
function isStepAligned(value, step, min = 0) {
  step = sanitizePositiveNumber(step);
  const roundedValue = Math.round((value - min) / step) * step + min;
  const tolerance = step * 1e-9;
  return Math.abs(value - roundedValue) < tolerance;
}
function addError(errorType, errorId, ref, dispatch) {
  dispatch?.({
    type: import_formfield_messages_state_reducer.FormFieldMessagesAction.ADDMESSAGE,
    payload: {
      id: errorId,
      message: (0, import_useFormFieldValidation.getI18nErrorMessage)(errorType, ref),
      validityStateKey: errorType,
      messageType: "formcontrol",
      variant: "error"
    }
  });
}
function getErrorKey(value, min, max, step) {
  const parsed = (0, import_format.parseNumber)(value);
  if (min !== void 0 && parsed < min) {
    return "rangeUnderflow";
  }
  if (max !== void 0 && parsed > max) {
    return "rangeOverflow";
  }
  if (step && step > 0 && !isStepAligned(parsed, step, min ?? 0)) {
    return "stepMismatch";
  }
  return null;
}
