// @dynatrace/strato-components/tables/DataTable/hooks/useColumns.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useColumns_exports = {};
__export(useColumns_exports, {
  isDataTableDisplayColumnDef: () => isDataTableDisplayColumnDef,
  isDataTableGroupColumnDef: () => isDataTableGroupColumnDef,
  useColumns: () => useColumns
});
module.exports = __toCommonJS(useColumns_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_makeAccessorFunction = require("./makeAccessorFunction.js");
var import_log_content_highlighting = require("../../utils/log-content-highlighting.js");
var import_replace_urls_with_links = require("../../utils/replace-urls-with-links.js");
var import_CellFormatter = require("../features/CellFormatter/CellFormatter.js");
var import_ColumnTypes = require("../features/ColumnTypes/ColumnTypes.js");
var import_Layout = require("../features/Layout/Layout.js");
var import_sort_types = require("../features/Sorting/sort-types.js");
function isDataTableGroupColumnDef(columnDef) {
  return columnDef !== null && typeof columnDef === "object" && "columns" in columnDef && Array.isArray(columnDef.columns);
}
function isDataTableDisplayColumnDef(columnDef) {
  return columnDef !== null && typeof columnDef === "object" && "accessor" in columnDef;
}
function splitAccessor(column) {
  if (typeof column.accessor === "string") {
    if (column.accessor.includes(".") || column.accessor.includes("[")) {
      return {
        accessorFn: (0, import_makeAccessorFunction.makeAccessorFunction)(column.accessor)
      };
    }
    return {
      accessorKey: column.accessor
    };
  }
  if (typeof column.accessor === "function") {
    return {
      accessorFn: column.accessor
    };
  }
  return {};
}
function createHeaderRenderer(header) {
  if (typeof header === "function") {
    return () => header();
  }
  return header;
}
function createCellRenderer(consumerCustomCell, consumerColumnDef) {
  return ({ getValue, row, column }) => {
    const lineWrap = column.getIsLineWrapped();
    return consumerCustomCell({
      value: getValue(),
      rowIndex: row.index,
      rowData: row.original,
      rowId: row.id,
      format: (value) => (0, import_CellFormatter.formatCellValue)(value, consumerColumnDef),
      formatLogContent: (logContent) => {
        const truncationLimit = (0, import_ColumnTypes.isDatatableLogContentColumnConfig)(
          consumerColumnDef.config
        ) ? consumerColumnDef.config.truncationLimit : void 0;
        return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_log_content_highlighting.LogContentHighlight,
          {
            logContent,
            lineWrap,
            truncationLimit
          }
        );
      },
      detectLinks: import_replace_urls_with_links.replaceUrlsWithLinks,
      isLineWrapped: lineWrap
    });
  };
}
function createDisplayColumnDef(columnDef) {
  const hasFormatter = columnDef.columnType !== "meterbar" && columnDef.columnType !== "sparkline" && columnDef.columnType !== "gantt" && !!columnDef.formatter;
  const composedDef = {
    id: columnDef.id,
    header: createHeaderRenderer(columnDef.header),
    label: columnDef.label,
    width: columnDef.width,
    minWidth: columnDef.minWidth,
    maxWidth: columnDef.maxWidth,
    alignment: columnDef.alignment ?? (0, import_ColumnTypes.getAlignmentBasedOnType)(columnDef.columnType),
    enableSorting: columnDef.disableSorting ? !columnDef.disableSorting : true,
    sortAccessor: columnDef.sortAccessor,
    sortDescFirst: columnDef.sortDescFirst,
    invertSorting: columnDef.sortInverted,
    enableHiding: !columnDef.disableColumnHiding,
    thresholds: columnDef.thresholds,
    meta: {
      config: columnDef.config,
      hasFormatter,
      hasCustomCellRenderer: columnDef.cell !== void 0,
      hasCustomHeaderRenderer: typeof columnDef.header !== "string"
    },
    ...splitAccessor(columnDef),
    ...(0, import_sort_types.applySortTypes)(columnDef)
  };
  if (columnDef.columnType) {
    composedDef.columnType = columnDef.columnType;
  }
  if (columnDef.cell) {
    const customCell = columnDef.cell;
    if (typeof customCell === "string") {
      composedDef.cell = customCell;
      return composedDef;
    }
    if (typeof customCell === "function") {
      return {
        ...composedDef,
        cell: createCellRenderer(customCell, columnDef)
      };
    }
    return composedDef;
  }
  return {
    ...composedDef,
    ...(0, import_ColumnTypes.applyColumnTypes)(columnDef, composedDef),
    ...(0, import_CellFormatter.applyColumnFormatters)(columnDef)
  };
}
function useColumns(propColumns) {
  return (0, import_react.useMemo)(() => {
    const createdColumns = [];
    const propColumnMap = /* @__PURE__ */ new Map();
    for (const firstLevelColumn of propColumns) {
      propColumnMap.set(firstLevelColumn.id, firstLevelColumn);
      if (isDataTableGroupColumnDef(firstLevelColumn)) {
        const groupColumn = {
          id: firstLevelColumn.id,
          header: createHeaderRenderer(firstLevelColumn.header),
          width: firstLevelColumn.width,
          enableSorting: false,
          columns: firstLevelColumn.columns?.map(
            (childColumn) => createDisplayColumnDef(childColumn)
          ),
          alignment: firstLevelColumn.alignment,
          meta: {
            hasCustomHeaderRenderer: typeof firstLevelColumn.header !== "string"
          }
        };
        createdColumns.push(groupColumn);
        for (const secondLevelColumn of firstLevelColumn.columns ?? []) {
          propColumnMap.set(
            secondLevelColumn.id,
            secondLevelColumn
          );
        }
      } else if (isDataTableDisplayColumnDef(firstLevelColumn)) {
        createdColumns.push(createDisplayColumnDef(firstLevelColumn));
      }
    }
    if (createdColumns.length > 0) {
      createdColumns.push(import_Layout.SPACER_COLUMN);
    }
    return [createdColumns, propColumnMap];
  }, [propColumns]);
}
