// @dynatrace/strato-components/tables/DataTable/hooks/useDataTableSlots.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useDataTableSlots_exports = {};
__export(useDataTableSlots_exports, {
  useDataTableSlots: () => useDataTableSlots
});
module.exports = __toCommonJS(useDataTableSlots_exports);
var import_react = require("react");
var import_react_is = require("react-is");
var import_DataTableEmptyState = require("../components/DataTableEmptyState.js");
var import_DataTableSelectedRowsActions = require("../components/TableActions/DataTableSelectedRowsActions.js");
var import_DataTableTableActions = require("../components/TableActions/DataTableTableActions.js");
var import_DataTableToolbar = require("../components/Toolbar/DataTableToolbar.js");
var import_Pagination = require("../features/Pagination/Pagination.js");
function useDataTableSlots(children) {
  return (0, import_react.useMemo)(() => {
    const slots = {
      pagination: null,
      toolbarProps: null,
      selectedRowsActions: null,
      emptyState: null,
      tableActions: null
    };
    import_react.Children.forEach(children, (child) => {
      if ((0, import_react_is.isElement)(child)) {
        if ((0, import_DataTableToolbar._isDataTableToolbar)(child)) {
          slots.toolbarProps = child.props;
        }
        if ((0, import_DataTableTableActions._isTableActions)(child)) {
          slots.tableActions = child;
        }
        if ((0, import_DataTableEmptyState._isEmptyState)(child)) {
          slots.emptyState = child;
        }
        if ((0, import_Pagination._isDataTablePagination)(child)) {
          slots.pagination = child;
        }
        if ((0, import_DataTableSelectedRowsActions._isDataTableSelectedRowsActions)(child)) {
          slots.selectedRowsActions = child;
        }
      }
    });
    return slots;
  }, [children]);
}
