// @dynatrace/strato-components/tables/DataTable/hooks/useHeaderDragAndDrop.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useHeaderDragAndDrop_exports = {};
__export(useHeaderDragAndDrop_exports, {
  useHeaderCellDragAndDrop: () => useHeaderCellDragAndDrop
});
module.exports = __toCommonJS(useHeaderDragAndDrop_exports);
var import_sortable = require("@dnd-kit/sortable");
var import_react = require("react");
var import_column_order_dnd_utils = require("../features/ColumnOrder/column-order-dnd-utils.js");
function useHeaderCellDragAndDrop({
  headerId,
  column,
  columnIndex,
  columnDragAndDropMode
}) {
  const touchStart = (0, import_react.useRef)();
  const portalRef = (0, import_react.useRef)(null);
  const { setNodeRef, isDragging, attributes, listeners } = (0, import_sortable.useSortable)({
    id: headerId,
    data: {
      portalRef,
      column,
      columnIndex
    }
  });
  const dragAndDropProps = columnDragAndDropMode === "none" ? void 0 : {
    ...attributes,
    draggable: true,
    onDragStart: (event) => listeners?.onMouseDown(event),
    onTouchStart: (event) => {
      touchStart.current = event;
    },
    onTouchMove: (event) => {
      if ((0, import_column_order_dnd_utils.isDragMotion)(event, touchStart?.current)) {
        listeners?.onTouchStart(event);
      }
    }
  };
  return {
    setNodeRef,
    isDragging,
    portalRef,
    dragAndDropProps
  };
}
