// @dynatrace/strato-components/esm/content/code-snippet/CodeSnippet.js
import { jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import writeText from "copy-to-clipboard";
import {
  forwardRef,
  useCallback,
  useMemo,
  useRef,
  useState
} from "react";
import { useIntl } from "react-intl";
import Borders from "@dynatrace/strato-design-tokens/borders";
import { CopyIcon } from "@dynatrace/strato-icons";
import { CodeHighlight } from "./CodeHighlight.js";
import * as styles from "./CodeSnippet.sty.js";
import { Button } from "../../buttons/button/Button.js";
import { useAriaLabelingProps } from "../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useSafeForwardProps } from "../../core/hooks/useSafeForwardProps.js";
import { Tooltip } from "../../overlays/tooltip/Tooltip.js";
const COMPONENT_NAME = "CodeSnippet";
const CodeSnippet = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const {
    children,
    language,
    // eslint-disable-next-line @typescript-eslint/no-deprecated
    allowCopy: allowCopyProp,
    showCopyAction: showCopyActionProp,
    showLineNumbers = true,
    lineBreaks = false,
    maxHeight,
    onCopy,
    size = "default",
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    style: consumerStyle,
    className: consumerClassName,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
  const ariaLabelingProps = useAriaLabelingProps(restProps);
  const forwardProps = useSafeForwardProps(restProps, ariaLabelingProps);
  const showCopyAction = showCopyActionProp ?? allowCopyProp ?? true;
  const joinedChildren = useMemo(
    () => Array.isArray(children) ? children.join("") : children,
    [children]
  );
  const copyText = `${joinedChildren?.toString() ?? ""}`;
  const [hasCopied, setHasCopied] = useState(false);
  const [isTooltipOpen, setIsTooltipOpen] = useState(false);
  const intl = useIntl();
  const handleCopy = async () => {
    writeText(copyText, { format: "text/plain" });
    setHasCopied(true);
    setIsTooltipOpen(true);
    onCopy?.();
  };
  const copySelection = useCallback(
    (event) => {
      const selection = document.getSelection();
      if (selection === null) {
        return;
      }
      event.clipboardData.setData("text/plain", selection.toString());
      event.preventDefault();
    },
    []
  );
  const copyLabel = intl.formatMessage({
    defaultMessage: "Copy",
    id: "Yi8j4dchtpkwn/OB",
    description: "Shown for elements that trigger a copy to clipboard"
  });
  const copiedLabel = intl.formatMessage({
    defaultMessage: "Copied",
    id: "THNwSvfvDSXAUapi",
    description: "Shown after copying to clipboard"
  });
  const message = useMemo(() => {
    return hasCopied ? copiedLabel : copyLabel;
  }, [hasCopied, copiedLabel, copyLabel]);
  const codeHighlightRef = useRef(null);
  return /* @__PURE__ */ jsxs(
    "div",
    {
      ref: forwardedRef,
      onCopy: copySelection,
      className: clsx(consumerClassName, styles.wrapper),
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...ariaLabelingProps,
      ...behavioralTrackingProps,
      ...forwardProps,
      "data-dt-component": COMPONENT_NAME,
      children: [
        /* @__PURE__ */ jsx(
          CodeHighlight,
          {
            ref: codeHighlightRef,
            language,
            code: copyText,
            showLineNumbers,
            lineBreaks,
            maxHeight,
            size
          }
        ),
        showCopyAction ? /* @__PURE__ */ jsx("div", { className: styles.copyButtonContainer({ size }), children: /* @__PURE__ */ jsx(
          Tooltip,
          {
            text: message,
            placement: "top",
            open: isTooltipOpen,
            delay: "none",
            onOpenChange: (isOpen) => {
              setIsTooltipOpen(isOpen);
              if (!isOpen) {
                setHasCopied(false);
              }
            },
            children: /* @__PURE__ */ jsx(
              Button,
              {
                variant: "default",
                size,
                "aria-label": copyLabel,
                onClick: handleCopy,
                style: { borderRadius: Borders.Radius.Field.Subdued },
                "data-dt-name": "Copy",
                children: /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(CopyIcon, {}) })
              }
            )
          }
        ) }) : null
      ]
    }
  );
});
CodeSnippet.displayName = COMPONENT_NAME;
export {
  CodeSnippet
};
