// @dynatrace/strato-components/esm/content/code-snippet/CodeHighlight.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import { forwardRef, useMemo } from "react";
import { ErrorBoundary } from "react-error-boundary";
import * as styles from "./CodeHighlight.sty.js";
import {
  HighlightExtended
} from "./HighlightExtended.js";
import { splitCodeForHighlighting } from "./utils/highlighting-utils.js";
import { getStratoLogger } from "../../core/sdk/sdk-logger.js";
function NonHighlightedBlock({
  showLineNumbers,
  size,
  isMultiLine,
  lineNumbers,
  code
}) {
  return /* @__PURE__ */ jsxs("div", { className: styles.line, children: [
    showLineNumbers ? /* @__PURE__ */ jsx(
      "span",
      {
        className: clsx(
          "line-number",
          styles.lineNumberContainer({ multiLine: isMultiLine })
        ),
        children: /* @__PURE__ */ jsx(
          "span",
          {
            className: styles.lineNumber({
              size,
              multiLine: isMultiLine
            }),
            children: lineNumbers
          }
        )
      }
    ) : null,
    /* @__PURE__ */ jsx(
      "span",
      {
        className: styles.lineContent({
          size,
          multiLine: isMultiLine
        }),
        children: code
      }
    )
  ] });
}
const CodeHighlight = /* @__PURE__ */ forwardRef((props, forwardedRef) => {
  const {
    code,
    language,
    showLineNumbers = true,
    lineBreaks = false,
    maxHeight,
    size = "default"
  } = props;
  const stratoLogger = getStratoLogger("CodeSnippet");
  const {
    highlightCode,
    noHighlightInline,
    noHighlightCode,
    noHighlightLineNumbers
  } = useMemo(() => splitCodeForHighlighting(code), [code]);
  const isNonHighlightedMultiLine = noHighlightLineNumbers.includes("\n");
  const hasNonHighlightedBlock = noHighlightCode.length > 0;
  const isFullyHighlighted = noHighlightInline.length === 0 && !hasNonHighlightedBlock;
  return /* @__PURE__ */ jsx(
    "pre",
    {
      className: styles.highlightWrapper({ size }),
      ref: forwardedRef,
      style: {
        whiteSpace: lineBreaks ? "pre-wrap" : "pre",
        maxHeight
      },
      children: /* @__PURE__ */ jsx(
        ErrorBoundary,
        {
          fallback: code,
          onError: (error, info) => {
            stratoLogger.error(error.message, error, {
              code,
              info
            });
          },
          children: /* @__PURE__ */ jsx(HighlightExtended, { language, code: highlightCode, children: ({ tokens, getLineProps, getTokenProps }) => {
            const isSingleLine = tokens.length === 1 && isFullyHighlighted;
            return /* @__PURE__ */ jsxs(Fragment, { children: [
              tokens.map((line, i) => {
                const isLastLine = i === tokens.length - 1;
                const { className: lineClassName, ...remainingLineProps } = getLineProps({ line });
                return /* @__PURE__ */ jsxs(
                  "div",
                  {
                    className: clsx(lineClassName, styles.line),
                    ...remainingLineProps,
                    children: [
                      showLineNumbers ? /* @__PURE__ */ jsx(
                        "span",
                        {
                          className: clsx(
                            "line-number",
                            styles.lineNumberContainer()
                          ),
                          children: /* @__PURE__ */ jsx(
                            "span",
                            {
                              className: styles.lineNumber({ size }),
                              "data-linenumber": i + 1
                            }
                          )
                        }
                      ) : null,
                      /* @__PURE__ */ jsxs(
                        "span",
                        {
                          className: styles.lineContent({ size }),
                          style: {
                            verticalAlign: isSingleLine ? "middle" : void 0
                          },
                          children: [
                            line.map((token, index) => /* @__PURE__ */ jsx("span", { ...getTokenProps({ token }) }, index)),
                            isLastLine && noHighlightInline.length > 0 ? noHighlightInline : null
                          ]
                        }
                      )
                    ]
                  },
                  i
                );
              }),
              hasNonHighlightedBlock ? /* @__PURE__ */ jsx(
                NonHighlightedBlock,
                {
                  showLineNumbers,
                  size,
                  isMultiLine: isNonHighlightedMultiLine,
                  lineNumbers: noHighlightLineNumbers,
                  code: noHighlightCode
                }
              ) : null
            ] });
          } })
        }
      )
    }
  );
});
CodeHighlight.displayName = "CodeHighlight";
export {
  CodeHighlight
};
