// @dynatrace/strato-components/esm/content/code-snippet/HighlightExtended.js
import { Fragment, jsx } from "react/jsx-runtime";
import { Prism, Highlight } from "prism-react-renderer";
import { useEffect, useState } from "react";
import { prismDqlLanguage } from "./prism-dql-language.js";
import { devLog } from "../../core/utils/logging.js";
import { DTPrismTheme } from "../../core/utils/prism-theme.js";
(typeof global !== "undefined" ? global : window).Prism = Prism;
Prism.languages.dql = prismDqlLanguage;
const SHORT_HEX_ID_TOKEN_NAME = "strato-short-hash";
const SHORT_HEX_ID_PATTERN = /\b[a-f0-9]{16}\b/i;
function enhancePrismLogLanguage() {
  const logLanguage = Prism.languages.log;
  if (!logLanguage || SHORT_HEX_ID_TOKEN_NAME in logLanguage) {
    return;
  }
  Prism.languages.insertBefore("log", "property", {
    [SHORT_HEX_ID_TOKEN_NAME]: {
      pattern: SHORT_HEX_ID_PATTERN,
      alias: "constant"
    }
  });
}
const additionalLanguageImports = {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  bash: () => import("prismjs/components/prism-bash.js"),
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  log: () => import("prismjs/components/prism-log.js"),
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  groovy: () => import("prismjs/components/prism-groovy.js")
};
const HighlightExtended = (props) => {
  const { language, code, children } = props;
  const [loadedLanguages, setLoadedLanguages] = useState(
    () => Object.fromEntries(
      Object.keys(additionalLanguageImports).map((lang) => [
        lang,
        Prism.languages[lang] !== void 0
      ])
    )
  );
  useEffect(() => {
    if (language && additionalLanguageImports[language] && !loadedLanguages[language]) {
      additionalLanguageImports[language]().then(() => {
        if (language === "log") {
          enhancePrismLogLanguage();
        }
        setLoadedLanguages((prev) => ({ ...prev, [language]: true }));
      }).catch((error) => {
        devLog(`Failed to load language '${language}':`, error);
      });
    }
  }, [language, loadedLanguages]);
  useEffect(() => {
    if (language === "log" && loadedLanguages.log) {
      enhancePrismLogLanguage();
    }
  }, [language, loadedLanguages.log]);
  return /* @__PURE__ */ jsx(
    Highlight,
    {
      language: language ?? "plain",
      code,
      theme: language === "log" ? DTPrismTheme.Log : DTPrismTheme.Default,
      Prism,
      children: ({ tokens, getLineProps, getTokenProps, className, style }) => /* @__PURE__ */ jsx(Fragment, { children: children({
        tokens,
        getLineProps,
        getTokenProps,
        className,
        style
      }) })
    }
  );
};
export {
  HighlightExtended
};
