// @dynatrace/strato-components/esm/content/code-snippet/prism-dql-language.js
import {
  booleanLiteralPattern,
  dqlUnitSuffixPattern,
  nullLiteralPattern,
  parameterNamePattern,
  sortedDqlCommands
} from "../../core/dql/dql-common.js";
const singleLineCommentPattern = /\/\/[^\n\r\v\f]*/;
const multiLineCommentPattern = /\/\*[\s\S]*?\*\//m;
const comments = {
  comment: {
    pattern: singleLineCommentPattern,
    greedy: true
  },
  commentMultiline: {
    pattern: multiLineCommentPattern,
    alias: "comment",
    greedy: true
  }
};
const stringPattern = /"(?:[^"\\]|\\.)*"/m;
const backtickStringPattern = /`(?:[^`\\]|\\.)*`/m;
const tripleQuoteStringPattern = /""""?"?(?:[^"]"?"?)*"""/m;
const dqlCommandNamePattern = new RegExp(
  `\\b(?:${sortedDqlCommands.join("|")})\\b`,
  "i"
);
const commandNamePattern = /[A-Za-z]+/;
const commandParametersPattern = new RegExp(
  `(${commandNamePattern.source})(?:${tripleQuoteStringPattern.source}|${backtickStringPattern.source}|${stringPattern.source}|${singleLineCommentPattern.source}|${multiLineCommentPattern.source}|[^|])*`
);
const commandPattern = new RegExp(
  `${commandNamePattern.source}${commandParametersPattern.source}\\|?`,
  "m"
);
const commandName = {
  commandName: {
    pattern: dqlCommandNamePattern,
    alias: "keyword"
  }
};
const prismDqlLanguage = {
  ...comments,
  /**
   * DQL language is defined as commands separated by the optional character (|).
   * Prism first matches an entire command, then matches the command parameters and command name
   * 'inside' the command.
   * Within the command parameters, Prism matches strings, variables, and numbers.
   *
   * Take great care of adding or modifying the regex patterns as this
   * may introduce recursion issues, due to the existence of three levels of regex matching.
   */
  command: {
    pattern: commandPattern,
    greedy: true,
    inside: {
      commandParameters: {
        pattern: commandParametersPattern,
        lookbehind: true,
        inside: {
          tripleQuoteString: {
            pattern: tripleQuoteStringPattern,
            alias: "string"
          },
          string: {
            pattern: stringPattern
            // Support strings surrounded by double quotes and include possible escaped quotes within
          },
          backtickstring: {
            pattern: backtickStringPattern,
            alias: "string"
          },
          number: new RegExp(`\\b\\d+(?:\\.\\d+)?${dqlUnitSuffixPattern}?\\b`),
          // Match numbers with optional unit suffix
          boolean: booleanLiteralPattern,
          // Match boolean literals (true/false)
          null: { pattern: nullLiteralPattern, alias: "boolean" },
          // null literal, styled as boolean
          parameterName: { pattern: parameterNamePattern, alias: "variable" },
          ...comments,
          commandName: {
            pattern: new RegExp(
              `(\\[\\s*)${dqlCommandNamePattern.source}`,
              "i"
            ),
            lookbehind: true,
            alias: "keyword"
          }
        },
        ...commandName
      },
      ...commandName
    }
  }
};
export {
  prismDqlLanguage
};
