// @dynatrace-sdk/client-platform-management-service/cjs/index.js
/**
 * @license
 * Copyright 2023-2025 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/client/platform-management-service/src/index.ts
var src_exports = {};
__export(src_exports, {
  ApiClientError: () => import_shared_errors6.ApiClientError,
  ApiGatewayError: () => import_shared_errors6.ApiGatewayError,
  BadRequest: () => BadRequest,
  ClientRequestError: () => import_shared_errors6.ClientRequestError,
  DefaultErrorResponse: () => DefaultErrorResponse,
  EffectivePermissionGranted: () => EffectivePermissionGranted,
  EffectivePermissionsClient: () => EffectivePermissionsClient,
  EnvironmentInfoType: () => EnvironmentInfoType,
  EnvironmentInformationClient: () => EnvironmentInformationClient,
  EnvironmentSettingsClient: () => EnvironmentSettingsClient,
  EnvironmentState: () => EnvironmentState,
  ErrorEnvelopeError: () => ErrorEnvelopeError,
  InvalidResponseError: () => import_shared_errors6.InvalidResponseError,
  LicenseInformationClient: () => LicenseInformationClient,
  NotFound: () => NotFound,
  ServiceUnavailable: () => ServiceUnavailable,
  Unauthorized: () => Unauthorized,
  effectivePermissionsClient: () => effectivePermissionsClient,
  environmentInformationClient: () => environmentInformationClient,
  environmentSettingsClient: () => environmentSettingsClient,
  isApiClientError: () => import_shared_errors6.isApiClientError,
  isApiGatewayError: () => import_shared_errors6.isApiGatewayError,
  isBadRequest: () => isBadRequest,
  isClientRequestError: () => import_shared_errors6.isClientRequestError,
  isDefaultErrorResponse: () => isDefaultErrorResponse,
  isErrorEnvelopeError: () => isErrorEnvelopeError,
  isInvalidResponseError: () => import_shared_errors6.isInvalidResponseError,
  isNotFound: () => isNotFound,
  isServiceUnavailable: () => isServiceUnavailable,
  isUnauthorized: () => isUnauthorized,
  licenseInformationClient: () => licenseInformationClient
});
module.exports = __toCommonJS(src_exports);

// packages/client/platform-management-service/src/lib/apis/effective-permissions-api.ts
var import_http_client = require("@dynatrace-sdk/http-client");
var import_shared_client_utils = require("@dynatrace-sdk/shared-client-utils");
var import_shared_errors2 = require("@dynatrace-sdk/shared-errors");

// packages/client/platform-management-service/src/lib/error-envelopes/error-envelope-error.ts
var import_shared_errors = require("@dynatrace-sdk/shared-errors");
var ErrorEnvelopeError = class extends import_shared_errors.ClientRequestError {
  isErrorEnvelopeError = true;
};
function isErrorEnvelopeError(e) {
  return e?.isApiClientError === true && e?.isClientRequestError === true && e?.isErrorEnvelopeError === true && e instanceof Error;
}

// packages/client/platform-management-service/src/lib/error-envelopes/responses/default-error-response.ts
var DefaultErrorResponse = class extends ErrorEnvelopeError {
  isDefaultErrorResponse = true;
  name = "DefaultErrorResponse";
};
function isDefaultErrorResponse(e) {
  return e?.isApiClientError === true && e?.isClientRequestError === true && e?.isErrorEnvelopeError === true && e?.isDefaultErrorResponse === true && e instanceof Error;
}

// packages/client/platform-management-service/src/lib/apis/effective-permissions-api.ts
var EffectivePermissionsClient = class {
  httpClient;
  shouldTransformDates = true;
  /**
   * @param {HttpClient} httpClientImplementation - You can provide custom http client as a parameter to constructor of a EffectivePermissionsClient. Custom http client should implement {HttpClient} interface.
   * @param {ApiClientOptions} [options] - Optional configuration for SDK Client.
   * @example
   * const effectivePermissionsClientNodeJs = new EffectivePermissionsClient(yourCustomImplementation);
   */
  constructor(httpClientImplementation, options) {
    this.httpClient = httpClientImplementation;
    this.shouldTransformDates = options?.transformDates ?? true;
  }
  /**
   * Get the effective permissions of the calling user in the environment
   *
   * **One of the following scopes is required:**
   * + app-engine:apps:run
   * + app-engine:functions:run
   *
   * <br/><br/>This endpoint resolves whether the caller (based on the provided OAuth token when calling this endpoint) has the requested permissions (see [IAM policy reference](https://dt-url.net/ut02qdv) for a list of permissions and their conditions). It resolves the permission via IAM and uses the optional context to evaluate conditional results. The provided context consists of key-value pairs where the `key` must correspond to the name of a supported condition of the requested permission and the `value` is the value that it this condition is evaluated against given a user's policies.
   *
   * For each requested permission the response indicates whether the user has the requested permission (`"granted": "true"`), does not have it (`"granted": "false"`) or only has it conditionally (`"granted": "condition"`), meaning that the provided values for conditions in the context are not sufficient to decide whether the caller has the permission.
   *
   * For environment API v2 scopes (i.e. scopes starting with environment-api:) roles on the environment and management zone level will also be considered.
   *
   * > ℹ️ **Settings permission** requests for scopes `settings:schemas:read`, `settings:objects:read`, and `settings:objects:write` will always result in a result with `"granted": "condition"`. You can use the dedicated [effective permissions endpoint for settings](/platform/swagger-ui/index.html?urls.primaryName=Classic+Environment+API+v2+%28for+SDK+generation%29) instead.
   *
   * > ℹ️ Apps must list all permissions they check in the list of `scopes` in their app configuration.
   *
   * You can learn more about querying user permissions on [Dynatrace Developer](https://dt-url.net/query-user-permissions).
   *
   * @returns {Promise<EffectivePermissions>} Success
   *
   * @example <caption>Code example</caption>
   * import { effectivePermissionsClient } from "@dynatrace-sdk/client-platform-management-service";
   *
   * const data =
   *   await effectivePermissionsClient.resolveEffectivePermissions(
   *     {
   *       body: {
   *         permissions: [
   *           { permission: "state:app-states:write" },
   *         ],
   *       },
   *     },
   *   );
   *
   * @throws {DefaultErrorResponse} Unexpected error
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async resolveEffectivePermissions(config) {
    if (!config) {
      throw new import_shared_errors2.ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const encodedBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformRequest)(config.body, []) : config.body;
    try {
      const response = await this.httpClient.send({
        url: `/platform/management/v1/effective-permissions:resolve`,
        method: "POST",
        requestBodyType: "json",
        body: encodedBody,
        headers: {
          "Content-Type": "application/json",
          Accept: "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
      } catch (err) {
        throw new import_shared_errors2.InvalidResponseError(
          `EffectivePermissionsClient.resolveEffectivePermissions:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if ((0, import_http_client.isHttpClientAbortError)(e) || (0, import_shared_errors2.isInvalidResponseError)(e)) {
        throw e;
      }
      if (!(0, import_http_client.isHttpClientResponseError)(e)) {
        throw new import_shared_errors2.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await (0, import_shared_client_utils.apiGatewayErrorsHandler)(response);
      const responseValue = await response.body("json");
      try {
        const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils.transformResponse)(responseValue, []) : responseValue;
        throw new DefaultErrorResponse(`err`, response, errorBody, (0, import_shared_client_utils.getErrorMessage)(errorBody, "Unexpected error"), e);
      } catch (err) {
        if (err instanceof DefaultErrorResponse) {
          throw err;
        }
        throw new import_shared_errors2.InvalidResponseError(
          `EffectivePermissionsClient.resolveEffectivePermissions:err`,
          err,
          responseValue,
          "ErrorEnvelope",
          void 0,
          response
        );
      }
    }
  }
};
var effectivePermissionsClient = /* @__PURE__ */ new EffectivePermissionsClient(import_http_client.httpClient);

// packages/client/platform-management-service/src/lib/apis/environment-information-api.ts
var import_http_client2 = require("@dynatrace-sdk/http-client");
var import_shared_client_utils2 = require("@dynatrace-sdk/shared-client-utils");
var import_shared_errors3 = require("@dynatrace-sdk/shared-errors");

// packages/client/platform-management-service/src/lib/error-envelopes/responses/bad-request.ts
var BadRequest = class extends ErrorEnvelopeError {
  isBadRequest = true;
  name = "BadRequest";
};
function isBadRequest(e) {
  return e?.isApiClientError === true && e?.isClientRequestError === true && e?.isErrorEnvelopeError === true && e?.isBadRequest === true && e instanceof Error;
}

// packages/client/platform-management-service/src/lib/error-envelopes/responses/not-found.ts
var NotFound = class extends ErrorEnvelopeError {
  isNotFound = true;
  name = "NotFound";
};
function isNotFound(e) {
  return e?.isApiClientError === true && e?.isClientRequestError === true && e?.isErrorEnvelopeError === true && e?.isNotFound === true && e instanceof Error;
}

// packages/client/platform-management-service/src/lib/error-envelopes/responses/service-unavailable.ts
var ServiceUnavailable = class extends ErrorEnvelopeError {
  isServiceUnavailable = true;
  name = "ServiceUnavailable";
};
function isServiceUnavailable(e) {
  return e?.isApiClientError === true && e?.isClientRequestError === true && e?.isErrorEnvelopeError === true && e?.isServiceUnavailable === true && e instanceof Error;
}

// packages/client/platform-management-service/src/lib/error-envelopes/responses/unauthorized.ts
var Unauthorized = class extends ErrorEnvelopeError {
  isUnauthorized = true;
  name = "Unauthorized";
};
function isUnauthorized(e) {
  return e?.isApiClientError === true && e?.isClientRequestError === true && e?.isErrorEnvelopeError === true && e?.isUnauthorized === true && e instanceof Error;
}

// packages/client/platform-management-service/src/lib/apis/environment-information-api.ts
var EnvironmentInformationClient = class {
  httpClient;
  shouldTransformDates = true;
  /**
   * @param {HttpClient} httpClientImplementation - You can provide custom http client as a parameter to constructor of a EnvironmentInformationClient. Custom http client should implement {HttpClient} interface.
   * @param {ApiClientOptions} [options] - Optional configuration for SDK Client.
   * @example
   * const environmentInformationClientNodeJs = new EnvironmentInformationClient(yourCustomImplementation);
   */
  constructor(httpClientImplementation, options) {
    this.httpClient = httpClientImplementation;
    this.shouldTransformDates = options?.transformDates ?? true;
  }
  /**
   * Get basic environment information
   *
   * **One of the following scopes is required:**
   * + app-engine:apps:run
   * + app-engine:functions:run
   *
   * <br/><br/>Get basic information about the current environment.
   *
   * @returns {Promise<EnvironmentInfo>} Info about the requested environment.
   *
   * @example <caption>Code example</caption>
   * import { environmentInformationClient } from "@dynatrace-sdk/client-platform-management-service";
   *
   * const data =
   *   await environmentInformationClient.getEnvironmentInformation();
   *
   * @throws {BadRequest} Bad request
   * @throws {Unauthorized} Authentication failed
   * @throws {NotFound} Resource not found
   * @throws {ServiceUnavailable} There is a temporary problem in the backend.
   * @throws {DefaultErrorResponse} Unexpected error
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async getEnvironmentInformation(config = {}) {
    try {
      const response = await this.httpClient.send({
        url: `/platform/management/v1/environment`,
        method: "GET",
        headers: {
          Accept: "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? (0, import_shared_client_utils2.transformResponse)(responseValue, [
          ["createTime", { datetime: true }],
          ["blockTime", { datetime: true }]
        ]) : responseValue;
      } catch (err) {
        throw new import_shared_errors3.InvalidResponseError(
          `EnvironmentInformationClient.getEnvironmentInformation:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if ((0, import_http_client2.isHttpClientAbortError)(e) || (0, import_shared_errors3.isInvalidResponseError)(e)) {
        throw e;
      }
      if (!(0, import_http_client2.isHttpClientResponseError)(e)) {
        throw new import_shared_errors3.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await (0, import_shared_client_utils2.apiGatewayErrorsHandler)(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils2.transformResponse)(responseValue, []) : responseValue;
            throw new BadRequest(`400`, response, errorBody, (0, import_shared_client_utils2.getErrorMessage)(errorBody, "Bad request"), e);
          } catch (err) {
            if (err instanceof BadRequest) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `EnvironmentInformationClient.getEnvironmentInformation:400`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 401: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils2.transformResponse)(responseValue, []) : responseValue;
            throw new Unauthorized(`401`, response, errorBody, (0, import_shared_client_utils2.getErrorMessage)(errorBody, "Authentication failed"), e);
          } catch (err) {
            if (err instanceof Unauthorized) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `EnvironmentInformationClient.getEnvironmentInformation:401`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils2.transformResponse)(responseValue, []) : responseValue;
            throw new NotFound(`404`, response, errorBody, (0, import_shared_client_utils2.getErrorMessage)(errorBody, "Resource not found"), e);
          } catch (err) {
            if (err instanceof NotFound) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `EnvironmentInformationClient.getEnvironmentInformation:404`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 503: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils2.transformResponse)(responseValue, []) : responseValue;
            throw new ServiceUnavailable(
              `503`,
              response,
              errorBody,
              (0, import_shared_client_utils2.getErrorMessage)(errorBody, "There is a temporary problem in the backend."),
              e
            );
          } catch (err) {
            if (err instanceof ServiceUnavailable) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `EnvironmentInformationClient.getEnvironmentInformation:503`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        default: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils2.transformResponse)(responseValue, []) : responseValue;
            throw new DefaultErrorResponse(
              `err`,
              response,
              errorBody,
              (0, import_shared_client_utils2.getErrorMessage)(errorBody, "Unexpected error"),
              e
            );
          } catch (err) {
            if (err instanceof DefaultErrorResponse) {
              throw err;
            }
            throw new import_shared_errors3.InvalidResponseError(
              `EnvironmentInformationClient.getEnvironmentInformation:err`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
      }
    }
  }
};
var environmentInformationClient = /* @__PURE__ */ new EnvironmentInformationClient(import_http_client2.httpClient);

// packages/client/platform-management-service/src/lib/apis/environment-settings-api.ts
var import_http_client3 = require("@dynatrace-sdk/http-client");
var import_shared_client_utils3 = require("@dynatrace-sdk/shared-client-utils");
var import_shared_errors4 = require("@dynatrace-sdk/shared-errors");
var EnvironmentSettingsClient = class {
  httpClient;
  shouldTransformDates = true;
  /**
   * @param {HttpClient} httpClientImplementation - You can provide custom http client as a parameter to constructor of a EnvironmentSettingsClient. Custom http client should implement {HttpClient} interface.
   * @param {ApiClientOptions} [options] - Optional configuration for SDK Client.
   * @example
   * const environmentSettingsClientNodeJs = new EnvironmentSettingsClient(yourCustomImplementation);
   */
  constructor(httpClientImplementation, options) {
    this.httpClient = httpClientImplementation;
    this.shouldTransformDates = options?.transformDates ?? true;
  }
  /**
   * Get settings for the environment
   *
   * **One of the following scopes is required:**
   * + app-engine:apps:run
   * + app-engine:functions:run
   *
   * <br/><br/>Gets settings for environment.
   *
   * @returns {Promise<SettingsResponse>} The settings for the requested environment
   *
   * @example <caption>Code example</caption>
   * import { environmentSettingsClient } from "@dynatrace-sdk/client-platform-management-service";
   *
   * const data =
   *   await environmentSettingsClient.getEnvironmentSettings();
   *
   * @throws {BadRequest} Bad request
   * @throws {Unauthorized} Authentication failed
   * @throws {NotFound} Resource not found
   * @throws {ServiceUnavailable} There is a temporary problem in the backend.
   * @throws {DefaultErrorResponse} Unexpected error
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async getEnvironmentSettings(config = {}) {
    try {
      const response = await this.httpClient.send({
        url: `/platform/management/v1/environment/settings`,
        method: "GET",
        headers: {
          Accept: "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? (0, import_shared_client_utils3.transformResponse)(responseValue, [["activeTrialSubscriptionEndDate", { datetime: true }]]) : responseValue;
      } catch (err) {
        throw new import_shared_errors4.InvalidResponseError(
          `EnvironmentSettingsClient.getEnvironmentSettings:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if ((0, import_http_client3.isHttpClientAbortError)(e) || (0, import_shared_errors4.isInvalidResponseError)(e)) {
        throw e;
      }
      if (!(0, import_http_client3.isHttpClientResponseError)(e)) {
        throw new import_shared_errors4.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await (0, import_shared_client_utils3.apiGatewayErrorsHandler)(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils3.transformResponse)(responseValue, []) : responseValue;
            throw new BadRequest(`400`, response, errorBody, (0, import_shared_client_utils3.getErrorMessage)(errorBody, "Bad request"), e);
          } catch (err) {
            if (err instanceof BadRequest) {
              throw err;
            }
            throw new import_shared_errors4.InvalidResponseError(
              `EnvironmentSettingsClient.getEnvironmentSettings:400`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 401: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils3.transformResponse)(responseValue, []) : responseValue;
            throw new Unauthorized(`401`, response, errorBody, (0, import_shared_client_utils3.getErrorMessage)(errorBody, "Authentication failed"), e);
          } catch (err) {
            if (err instanceof Unauthorized) {
              throw err;
            }
            throw new import_shared_errors4.InvalidResponseError(
              `EnvironmentSettingsClient.getEnvironmentSettings:401`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils3.transformResponse)(responseValue, []) : responseValue;
            throw new NotFound(`404`, response, errorBody, (0, import_shared_client_utils3.getErrorMessage)(errorBody, "Resource not found"), e);
          } catch (err) {
            if (err instanceof NotFound) {
              throw err;
            }
            throw new import_shared_errors4.InvalidResponseError(
              `EnvironmentSettingsClient.getEnvironmentSettings:404`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 503: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils3.transformResponse)(responseValue, []) : responseValue;
            throw new ServiceUnavailable(
              `503`,
              response,
              errorBody,
              (0, import_shared_client_utils3.getErrorMessage)(errorBody, "There is a temporary problem in the backend."),
              e
            );
          } catch (err) {
            if (err instanceof ServiceUnavailable) {
              throw err;
            }
            throw new import_shared_errors4.InvalidResponseError(
              `EnvironmentSettingsClient.getEnvironmentSettings:503`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        default: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils3.transformResponse)(responseValue, []) : responseValue;
            throw new DefaultErrorResponse(
              `err`,
              response,
              errorBody,
              (0, import_shared_client_utils3.getErrorMessage)(errorBody, "Unexpected error"),
              e
            );
          } catch (err) {
            if (err instanceof DefaultErrorResponse) {
              throw err;
            }
            throw new import_shared_errors4.InvalidResponseError(
              `EnvironmentSettingsClient.getEnvironmentSettings:err`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
      }
    }
  }
};
var environmentSettingsClient = /* @__PURE__ */ new EnvironmentSettingsClient(import_http_client3.httpClient);

// packages/client/platform-management-service/src/lib/apis/license-information-api.ts
var import_http_client4 = require("@dynatrace-sdk/http-client");
var import_shared_client_utils4 = require("@dynatrace-sdk/shared-client-utils");
var import_shared_errors5 = require("@dynatrace-sdk/shared-errors");
var LicenseInformationClient = class {
  httpClient;
  shouldTransformDates = true;
  /**
   * @param {HttpClient} httpClientImplementation - You can provide custom http client as a parameter to constructor of a LicenseInformationClient. Custom http client should implement {HttpClient} interface.
   * @param {ApiClientOptions} [options] - Optional configuration for SDK Client.
   * @example
   * const licenseInformationClientNodeJs = new LicenseInformationClient(yourCustomImplementation);
   */
  constructor(httpClientImplementation, options) {
    this.httpClient = httpClientImplementation;
    this.shouldTransformDates = options?.transformDates ?? true;
  }
  /**
   * Get basic license information
   *
   * **One of the following scopes is required:**
   * + app-engine:apps:run
   * + app-engine:functions:run
   *
   * <br/><br/>Get basic license information about the current environment.
   *
   * @returns {Promise<License>} The license info of the requested environment.
   *
   * @example <caption>Code example</caption>
   * import { licenseInformationClient } from "@dynatrace-sdk/client-platform-management-service";
   *
   * const data = await licenseInformationClient.getLicense();
   *
   * @throws {BadRequest} Bad request
   * @throws {Unauthorized} Authentication failed
   * @throws {NotFound} Resource not found
   * @throws {ServiceUnavailable} There is a temporary problem in the backend.
   * @throws {DefaultErrorResponse} Unexpected error
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async getLicense(config = {}) {
    try {
      const response = await this.httpClient.send({
        url: `/platform/management/v1/environment/license`,
        method: "GET",
        headers: {
          Accept: "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? (0, import_shared_client_utils4.transformResponse)(responseValue, []) : responseValue;
      } catch (err) {
        throw new import_shared_errors5.InvalidResponseError(
          `LicenseInformationClient.getLicense:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if ((0, import_http_client4.isHttpClientAbortError)(e) || (0, import_shared_errors5.isInvalidResponseError)(e)) {
        throw e;
      }
      if (!(0, import_http_client4.isHttpClientResponseError)(e)) {
        throw new import_shared_errors5.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await (0, import_shared_client_utils4.apiGatewayErrorsHandler)(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils4.transformResponse)(responseValue, []) : responseValue;
            throw new BadRequest(`400`, response, errorBody, (0, import_shared_client_utils4.getErrorMessage)(errorBody, "Bad request"), e);
          } catch (err) {
            if (err instanceof BadRequest) {
              throw err;
            }
            throw new import_shared_errors5.InvalidResponseError(
              `LicenseInformationClient.getLicense:400`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 401: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils4.transformResponse)(responseValue, []) : responseValue;
            throw new Unauthorized(`401`, response, errorBody, (0, import_shared_client_utils4.getErrorMessage)(errorBody, "Authentication failed"), e);
          } catch (err) {
            if (err instanceof Unauthorized) {
              throw err;
            }
            throw new import_shared_errors5.InvalidResponseError(
              `LicenseInformationClient.getLicense:401`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils4.transformResponse)(responseValue, []) : responseValue;
            throw new NotFound(`404`, response, errorBody, (0, import_shared_client_utils4.getErrorMessage)(errorBody, "Resource not found"), e);
          } catch (err) {
            if (err instanceof NotFound) {
              throw err;
            }
            throw new import_shared_errors5.InvalidResponseError(
              `LicenseInformationClient.getLicense:404`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 503: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils4.transformResponse)(responseValue, []) : responseValue;
            throw new ServiceUnavailable(
              `503`,
              response,
              errorBody,
              (0, import_shared_client_utils4.getErrorMessage)(errorBody, "There is a temporary problem in the backend."),
              e
            );
          } catch (err) {
            if (err instanceof ServiceUnavailable) {
              throw err;
            }
            throw new import_shared_errors5.InvalidResponseError(
              `LicenseInformationClient.getLicense:503`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        default: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils4.transformResponse)(responseValue, []) : responseValue;
            throw new DefaultErrorResponse(
              `err`,
              response,
              errorBody,
              (0, import_shared_client_utils4.getErrorMessage)(errorBody, "Unexpected error"),
              e
            );
          } catch (err) {
            if (err instanceof DefaultErrorResponse) {
              throw err;
            }
            throw new import_shared_errors5.InvalidResponseError(
              `LicenseInformationClient.getLicense:err`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
      }
    }
  }
  /**
   * Get basic license settings information
   *
   * **One of the following scopes is required:**
   * + app-engine:apps:run
   * + app-engine:functions:run
   *
   * <br/><br/>Get basic license settings information about the current environment
   *
   * @returns {Promise<LicenseSettingsResponse>} The license settings info of the requested environment.
   *
   * @example <caption>Code example</caption>
   * import { licenseInformationClient } from "@dynatrace-sdk/client-platform-management-service";
   *
   * const data =
   *   await licenseInformationClient.getLicenseSettings();
   *
   * @throws {BadRequest} Bad request
   * @throws {Unauthorized} Authentication failed
   * @throws {NotFound} Resource not found
   * @throws {ServiceUnavailable} There is a temporary problem in the backend.
   * @throws {DefaultErrorResponse} Unexpected error
   * @throws {InvalidResponseError}
   * @throws {ApiClientError}
   * @throws {ClientRequestError}
   */
  async getLicenseSettings(config = {}) {
    if (!config) {
      throw new import_shared_errors5.ApiClientError("API client error", "API client call is missing mandatory config parameter");
    }
    const query = (0, import_shared_client_utils4.toQueryString)({ keys: config.keys });
    try {
      const response = await this.httpClient.send({
        url: `/platform/management/v1/environment/license/settings${query}`,
        method: "GET",
        headers: {
          Accept: "application/json"
        },
        abortSignal: config instanceof EventTarget ? config : config.abortSignal,
        statusValidator: (status) => {
          return [200].includes(status);
        }
      });
      const responseValue = await response.body("json");
      try {
        return this.shouldTransformDates ? (0, import_shared_client_utils4.transformResponse)(responseValue, []) : responseValue;
      } catch (err) {
        throw new import_shared_errors5.InvalidResponseError(
          `LicenseInformationClient.getLicenseSettings:200`,
          err,
          responseValue,
          void 0,
          void 0,
          response
        );
      }
    } catch (e) {
      if ((0, import_http_client4.isHttpClientAbortError)(e) || (0, import_shared_errors5.isInvalidResponseError)(e)) {
        throw e;
      }
      if (!(0, import_http_client4.isHttpClientResponseError)(e)) {
        throw new import_shared_errors5.ApiClientError("UnexpectedError", "Unexpected error", e);
      }
      const response = e.response;
      await (0, import_shared_client_utils4.apiGatewayErrorsHandler)(response);
      switch (response.status) {
        case 400: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils4.transformResponse)(responseValue, []) : responseValue;
            throw new BadRequest(`400`, response, errorBody, (0, import_shared_client_utils4.getErrorMessage)(errorBody, "Bad request"), e);
          } catch (err) {
            if (err instanceof BadRequest) {
              throw err;
            }
            throw new import_shared_errors5.InvalidResponseError(
              `LicenseInformationClient.getLicenseSettings:400`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 401: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils4.transformResponse)(responseValue, []) : responseValue;
            throw new Unauthorized(`401`, response, errorBody, (0, import_shared_client_utils4.getErrorMessage)(errorBody, "Authentication failed"), e);
          } catch (err) {
            if (err instanceof Unauthorized) {
              throw err;
            }
            throw new import_shared_errors5.InvalidResponseError(
              `LicenseInformationClient.getLicenseSettings:401`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 404: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils4.transformResponse)(responseValue, []) : responseValue;
            throw new NotFound(`404`, response, errorBody, (0, import_shared_client_utils4.getErrorMessage)(errorBody, "Resource not found"), e);
          } catch (err) {
            if (err instanceof NotFound) {
              throw err;
            }
            throw new import_shared_errors5.InvalidResponseError(
              `LicenseInformationClient.getLicenseSettings:404`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        case 503: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils4.transformResponse)(responseValue, []) : responseValue;
            throw new ServiceUnavailable(
              `503`,
              response,
              errorBody,
              (0, import_shared_client_utils4.getErrorMessage)(errorBody, "There is a temporary problem in the backend."),
              e
            );
          } catch (err) {
            if (err instanceof ServiceUnavailable) {
              throw err;
            }
            throw new import_shared_errors5.InvalidResponseError(
              `LicenseInformationClient.getLicenseSettings:503`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
        default: {
          const responseValue = await response.body("json");
          try {
            const errorBody = this.shouldTransformDates ? (0, import_shared_client_utils4.transformResponse)(responseValue, []) : responseValue;
            throw new DefaultErrorResponse(
              `err`,
              response,
              errorBody,
              (0, import_shared_client_utils4.getErrorMessage)(errorBody, "Unexpected error"),
              e
            );
          } catch (err) {
            if (err instanceof DefaultErrorResponse) {
              throw err;
            }
            throw new import_shared_errors5.InvalidResponseError(
              `LicenseInformationClient.getLicenseSettings:err`,
              err,
              responseValue,
              "ErrorEnvelope",
              void 0,
              response
            );
          }
        }
      }
    }
  }
};
var licenseInformationClient = /* @__PURE__ */ new LicenseInformationClient(import_http_client4.httpClient);

// packages/client/platform-management-service/src/lib/error-envelopes/index.ts
var import_shared_errors6 = require("@dynatrace-sdk/shared-errors");

// packages/client/platform-management-service/src/lib/models/effective-permission-granted.ts
var EffectivePermissionGranted = /* @__PURE__ */ ((EffectivePermissionGranted2) => {
  EffectivePermissionGranted2["True"] = "true";
  EffectivePermissionGranted2["False"] = "false";
  EffectivePermissionGranted2["Condition"] = "condition";
  return EffectivePermissionGranted2;
})(EffectivePermissionGranted || {});

// packages/client/platform-management-service/src/lib/models/environment-info-type.ts
var EnvironmentInfoType = /* @__PURE__ */ ((EnvironmentInfoType2) => {
  EnvironmentInfoType2["Internal"] = "INTERNAL";
  EnvironmentInfoType2["Customer"] = "CUSTOMER";
  EnvironmentInfoType2["SelfMonitoring"] = "SELF_MONITORING";
  return EnvironmentInfoType2;
})(EnvironmentInfoType || {});

// packages/client/platform-management-service/src/lib/models/environment-state.ts
var EnvironmentState = /* @__PURE__ */ ((EnvironmentState2) => {
  EnvironmentState2["BeingCreated"] = "BEING_CREATED";
  EnvironmentState2["CreatedNotInitialized"] = "CREATED_NOT_INITIALIZED";
  EnvironmentState2["Active"] = "ACTIVE";
  EnvironmentState2["Deactivated"] = "DEACTIVATED";
  EnvironmentState2["BeingDeleted"] = "BEING_DELETED";
  EnvironmentState2["Blocked"] = "BLOCKED";
  EnvironmentState2["DeletionFailed"] = "DELETION_FAILED";
  EnvironmentState2["PreAllocated"] = "PRE_ALLOCATED";
  return EnvironmentState2;
})(EnvironmentState || {});
