// @dynatrace/strato-components/charts/meter-bar/shared/components/meter-bar-layout/hooks/useComputedSegmentSize.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useComputedSegmentSize_exports = {};
__export(useComputedSegmentSize_exports, {
  useComputedSegmentSize: () => useComputedSegmentSize
});
module.exports = __toCommonJS(useComputedSegmentSize_exports);
var import_react = require("react");
var import_useObserveContainerSize = require("../../../../../core/hooks/useObserveContainerSize.js");
var import_constants = require("../../../constants.js");
const useComputedSegmentSize = () => {
  const [segmentSize, setSegmentSize] = (0, import_react.useState)({
    width: 0,
    height: 0
  });
  const labelRef = (0, import_react.useRef)(null);
  const valueLabelRef = (0, import_react.useRef)(null);
  const minLabelRef = (0, import_react.useRef)(null);
  const maxLabelRef = (0, import_react.useRef)(null);
  const legendRef = (0, import_react.useRef)(null);
  const [{ wrapperHeight, width, height }, containerRef] = (0, import_useObserveContainerSize.useObserveContainerSize)();
  const getGridHeightSizes = () => ({
    labelArea: labelRef?.current?.clientHeight ?? 0,
    valueLabelArea: valueLabelRef?.current?.clientHeight ?? 0,
    minLabelArea: minLabelRef?.current?.clientHeight ?? 0,
    maxLabelArea: maxLabelRef?.current?.clientHeight ?? 0,
    legendArea: legendRef?.current?.clientHeight ?? 0
  });
  (0, import_react.useLayoutEffect)(() => {
    const {
      labelArea,
      valueLabelArea,
      minLabelArea,
      maxLabelArea,
      legendArea
    } = getGridHeightSizes();
    const firstRowHeight = Math.max(labelArea, valueLabelArea);
    const secondRowHeight = Math.max(minLabelArea, maxLabelArea);
    const chartHeight = Math.min(
      height,
      wrapperHeight - firstRowHeight - secondRowHeight - legendArea
    );
    const plotArea = Math.max(import_constants.SMALL_CHART_SIZE, chartHeight);
    setSegmentSize({ width, height: plotArea });
  }, [
    width,
    height,
    wrapperHeight,
    valueLabelRef,
    minLabelRef,
    maxLabelRef,
    legendRef,
    labelRef
  ]);
  return {
    segmentSize,
    labelRef,
    valueLabelRef,
    minLabelRef,
    maxLabelRef,
    plotAreaRef: containerRef,
    legendRef
  };
};
