// @dynatrace/strato-components/charts/core/components/legend/range/constants.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var constants_exports = {};
__export(constants_exports, {
  CONTAINER_PADDING: () => CONTAINER_PADDING,
  CSS_CONTAINER_PADDING: () => CSS_CONTAINER_PADDING,
  DEFAULT_CONTAINER_SIZE: () => DEFAULT_CONTAINER_SIZE,
  DEFAULT_FORMATTER: () => DEFAULT_FORMATTER,
  DEFAULT_GAP_THICKNESS: () => DEFAULT_GAP_THICKNESS,
  DEFAULT_ITEM_THICKNESS: () => DEFAULT_ITEM_THICKNESS,
  DEFAULT_LABEL_FONT_COLOR: () => DEFAULT_LABEL_FONT_COLOR,
  DEFAULT_LABEL_FONT_SIZE: () => DEFAULT_LABEL_FONT_SIZE,
  DEFAULT_SEGMENT_COLOR: () => DEFAULT_SEGMENT_COLOR,
  DEFAULT_SEGMENT_SIZE: () => DEFAULT_SEGMENT_SIZE,
  DEFAULT_TRUNCATION_MODE: () => DEFAULT_TRUNCATION_MODE,
  GAP_BETWEEN_LABELS: () => GAP_BETWEEN_LABELS,
  HOVERED_LABEL_FONT_WEIGHT: () => HOVERED_LABEL_FONT_WEIGHT,
  LABEL_ALIGNMENT: () => LABEL_ALIGNMENT,
  LABEL_FONT_COLOR_HOVERED: () => LABEL_FONT_COLOR_HOVERED,
  LABEL_FONT_WEIGHT: () => LABEL_FONT_WEIGHT,
  LABEL_GAP: () => LABEL_GAP,
  LABEL_HOVERED_PADDING: () => LABEL_HOVERED_PADDING,
  MAX_LEGEND_HEIGHT: () => MAX_LEGEND_HEIGHT,
  MAX_LEGEND_WIDTH: () => MAX_LEGEND_WIDTH,
  OVERLAY_ADDITIONAL_SIZE: () => OVERLAY_ADDITIONAL_SIZE,
  OVERLAY_OFFSET: () => OVERLAY_OFFSET,
  RANGE_LEGEND_CLASS: () => RANGE_LEGEND_CLASS,
  STROKE_WIDTH: () => STROKE_WIDTH,
  TEXT_ELLIPSIS: () => TEXT_ELLIPSIS,
  TRUNCATION_MODE: () => TRUNCATION_MODE
});
module.exports = __toCommonJS(constants_exports);
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
var import_spacings = __toESM(require("@dynatrace/strato-design-tokens/spacings"));
var import_typography = __toESM(require("@dynatrace/strato-design-tokens/typography"));
var import_units = require("@dynatrace-sdk/units");
const getNumberFromString = (value) => {
  const extractedNumber = value.match(/\d+/);
  return extractedNumber ? +extractedNumber[0] : 0;
};
const LABEL_ALIGNMENT = {
  Start: "start",
  Center: "center",
  End: "end"
};
const TRUNCATION_MODE = {
  End: "end",
  Middle: "middle",
  Start: "start"
};
const DEFAULT_TRUNCATION_MODE = TRUNCATION_MODE.Middle;
const DEFAULT_FORMATTER = (value) => (0, import_units.format)(value);
const DEFAULT_ITEM_THICKNESS = 10;
const DEFAULT_SEGMENT_SIZE = 6;
const DEFAULT_SEGMENT_COLOR = import_colors.default.Background.Field.Neutral.Disabled;
const DEFAULT_GAP_THICKNESS = 1;
const DEFAULT_LABEL_FONT_SIZE = import_typography.default.Text.Small.Default.Size;
const DEFAULT_LABEL_FONT_COLOR = import_colors.default.Text.Neutral.Subdued;
const LABEL_FONT_COLOR_HOVERED = import_colors.default.Text.Neutral.Default;
const LABEL_FONT_WEIGHT = import_typography.default.Text.Small.Default.Weight;
const HOVERED_LABEL_FONT_WEIGHT = import_typography.default.Text.Small.Emphasized.Weight;
const GAP_BETWEEN_LABELS = 8;
const TEXT_ELLIPSIS = "...";
const STROKE_WIDTH = 1;
const MAX_LEGEND_HEIGHT = 360;
const MAX_LEGEND_WIDTH = 360;
const CSS_CONTAINER_PADDING = import_spacings.default.Size8;
const LABEL_HOVERED_PADDING = getNumberFromString(import_spacings.default.Size8);
const LABEL_GAP = getNumberFromString(import_spacings.default.Size2);
const CONTAINER_PADDING = getNumberFromString(CSS_CONTAINER_PADDING);
const DEFAULT_CONTAINER_SIZE = DEFAULT_ITEM_THICKNESS + CONTAINER_PADDING * 2;
const RANGE_LEGEND_CLASS = "range-legend";
const OVERLAY_ADDITIONAL_SIZE = getNumberFromString(import_spacings.default.Size8);
const OVERLAY_OFFSET = OVERLAY_ADDITIONAL_SIZE / 2;
