// @dynatrace/strato-components/charts/core/components/legend/range/RangeLegend.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var RangeLegend_exports = {};
__export(RangeLegend_exports, {
  RangeLegend: () => RangeLegend
});
module.exports = __toCommonJS(RangeLegend_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_RangeLegendMarkers = require("./components/RangeLegendMarkers/RangeLegendMarkers.js");
var import_RangeLegendSegments = require("./components/RangeLegendSegments/RangeLegendSegments.js");
var import_constants = require("./constants.js");
var import_RangeLegendProviders = require("./providers/RangeLegendProviders.js");
var import_RangeLegend_css = require("./RangeLegend.sty.js");
var import_get_legend_scale = require("./utils/get-legend-scale.js");
var import_map_to_internal_range_items = require("./utils/map-to-internal-range-items.js");
var import_Flex = require("../../../../../layouts/flex/Flex.js");
var import_constants2 = require("../../../constants.js");
var import_useChartLayoutRatio = require("../../chart-layout/hooks/useChartLayoutRatio.js");
var import_LegendSizeContext = require("../shared/context/LegendSizeContext.js");
const RangeLegend = (props) => {
  const {
    items,
    orientation = import_constants2.ORIENTATION.Horizontal,
    formatter = import_constants.DEFAULT_FORMATTER,
    truncationMode = import_constants.DEFAULT_TRUNCATION_MODE,
    min,
    max
  } = props;
  const isVertical = orientation === import_constants2.ORIENTATION.Vertical;
  const { width, height } = (0, import_LegendSizeContext.useLegendSize)();
  const { legendRef } = (0, import_useChartLayoutRatio.useChartLayoutRatio)();
  const internalRangeItems = (0, import_map_to_internal_range_items.mapToInternalRangeItems)(items);
  if (!internalRangeItems.length) {
    return null;
  }
  const legendHeight = Math.max(height, import_constants.DEFAULT_CONTAINER_SIZE);
  const legendWidth = Math.max(width, import_constants.DEFAULT_CONTAINER_SIZE);
  const containerLegendHeight = legendHeight - import_constants.CONTAINER_PADDING;
  const containerLegendWidth = legendWidth - import_constants.CONTAINER_PADDING;
  const containerSize = isVertical ? Math.min(containerLegendHeight, import_constants.MAX_LEGEND_HEIGHT) : Math.min(containerLegendWidth, import_constants.MAX_LEGEND_WIDTH);
  const scale = (0, import_get_legend_scale.getLegendScale)(
    internalRangeItems,
    containerSize,
    isVertical,
    min,
    max
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_RangeLegendProviders.RangeLegendProviders,
    {
      formatter,
      orientation,
      scale,
      truncationMode,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        "div",
        {
          ref: legendRef,
          className: (0, import_clsx.default)((0, import_RangeLegend_css.legendContainer)({ isVertical }), import_constants.RANGE_LEGEND_CLASS),
          style: {
            width: `${legendWidth}px`,
            height: `${legendHeight}px`
          },
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
            import_Flex.Flex,
            {
              width: `${containerLegendWidth}px`,
              height: `${containerLegendHeight}px`,
              maxWidth: import_constants.MAX_LEGEND_WIDTH,
              maxHeight: import_constants.MAX_LEGEND_HEIGHT,
              style: { position: "relative" },
              children: [
                /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  import_RangeLegendSegments.RangeLegendSegments,
                  {
                    items: internalRangeItems,
                    maxValue: max,
                    minValue: min
                  }
                ),
                /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  import_RangeLegendMarkers.RangeLegendMarkers,
                  {
                    items: internalRangeItems,
                    maxValue: max,
                    minValue: min
                  }
                )
              ]
            }
          )
        }
      )
    }
  );
};
