// @dynatrace/strato-components/esm/charts/core/components/interactions-overlay/interactions/zoom/dragInteractions.js
import { centerPositionVertically } from "../../utils/center-position-vertically.js";
import { normalizeSelectionArea } from "../../utils/normalize-selection-area.js";
const zoomDragStart = ({
  event,
  onSelectStart
}) => {
  onSelectStart?.({
    triggeredBy: "pointer",
    position: event.startHoldingAt
  });
};
const zoomDragMove = ({
  event,
  height,
  width,
  onSelectChange
}) => {
  const {
    startHoldingAt: startHoldingAtProp,
    currentPosition: currentPositionProp
  } = event;
  const startHoldingAt = centerPositionVertically(startHoldingAtProp, height);
  const currentPosition = centerPositionVertically(currentPositionProp, height);
  const [start, end] = normalizeSelectionArea([
    startHoldingAt,
    currentPosition
  ]);
  if (start.relative.x < 0 || end.relative.x > width) {
    return;
  }
  onSelectChange?.({
    triggeredBy: "pointer",
    start,
    end,
    mode: "zoom"
  });
};
const zoomDragEnd = ({
  event,
  height,
  width,
  lastMousePositionRef,
  onSelectEnd,
  onZoomChange
}) => {
  const { startHoldingAt: startHoldingAtProp, releasedAt: releasedAtProp } = event;
  const startHoldingAt = centerPositionVertically(startHoldingAtProp, height);
  const releasedAt = centerPositionVertically(releasedAtProp, height);
  const [start, end] = normalizeSelectionArea([startHoldingAt, releasedAt]);
  if (start.relative.x < 0 || end.relative.x > width) {
    return;
  }
  const deltaStart = start.relative.x;
  const deltaEnd = end.relative.x - width;
  onSelectEnd?.({ cancelled: true });
  onZoomChange?.({
    triggeredBy: "pointer",
    deltaStart,
    deltaEnd,
    behavior: "zoom-x"
  });
  lastMousePositionRef.current = releasedAt;
};
export {
  zoomDragEnd,
  zoomDragMove,
  zoomDragStart
};
