// @dynatrace/strato-components/esm/charts/core/components/interactions-overlay/interactions/zoom/keyboardInteractions.js
import { isUndefined } from "lodash-es";
import { ZOOM_FACTOR } from "../../constants.js";
const zoomKeyboardEscape = ({
  selectionArea,
  eventCancelled,
  onSelectEnd,
  lastMousePositionRef,
  pinned,
  onHover,
  onTogglePin
}) => {
  if (!isUndefined(selectionArea)) {
    eventCancelled.current = true;
    onSelectEnd?.({ cancelled: true });
    if (lastMousePositionRef.current) {
      onHover?.({
        triggeredBy: "pointer",
        position: lastMousePositionRef.current
      });
    }
  }
  if (pinned) {
    onTogglePin?.({
      triggeredBy: "keyboard",
      pin: false,
      position: lastMousePositionRef.current
    });
  }
};
const zoomKeyboardEnter = ({
  selectionArea,
  lastMousePositionRef,
  width,
  onSelectEnd,
  onZoomChange,
  onHover
}) => {
  if (selectionArea) {
    const deltaStart = selectionArea[0].relative.x;
    const deltaEnd = selectionArea[1].relative.x - width;
    onSelectEnd?.({ cancelled: true });
    onZoomChange?.({
      triggeredBy: "keyboard",
      deltaStart,
      deltaEnd,
      behavior: "zoom-x"
    });
    if (lastMousePositionRef.current) {
      onHover?.({
        triggeredBy: "pointer",
        position: lastMousePositionRef.current
      });
    }
  }
};
const zoomKeyboardUp = ({
  width,
  onZoomChange
}) => {
  const delta = Math.floor(width * ZOOM_FACTOR / 2);
  onZoomChange?.({
    triggeredBy: "keyboard",
    deltaStart: delta,
    deltaEnd: -delta,
    behavior: "zoom-in"
  });
};
const zoomKeyboardDown = ({
  width,
  onZoomChange
}) => {
  const delta = Math.floor(width * ZOOM_FACTOR / 2);
  onZoomChange?.({
    triggeredBy: "keyboard",
    deltaStart: -delta,
    deltaEnd: delta,
    behavior: "zoom-out"
  });
};
export {
  zoomKeyboardDown,
  zoomKeyboardEnter,
  zoomKeyboardEscape,
  zoomKeyboardUp
};
