// @dynatrace/strato-components/charts/meter-bar/single-meter-bar/utils/build-meter-bar-chart-config.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var build_meter_bar_chart_config_exports = {};
__export(build_meter_bar_chart_config_exports, {
  buildMeterBarChartConfig: () => buildMeterBarChartConfig
});
module.exports = __toCommonJS(build_meter_bar_chart_config_exports);
var import_apply_threshold_defaults = require("./apply-threshold-defaults.js");
var import_ColorRule = require("../../../core/slots/ColorRule.js");
var import_constants = require("../../shared/constants.js");
var import_meter_bar_toolbar = require("../../shared/types/meter-bar-toolbar.js");
var import_default = require("../types/default.js");
const getTooltipConfig = (slots, consumerConfig) => {
  const consumerTooltip = consumerConfig.tooltip !== void 0 ? { hidden: !consumerConfig.tooltip } : void 0;
  return { ...import_default.defaultConfig.tooltip, ...consumerTooltip, ...slots.tooltip };
};
const getToolbarConfig = (toolbarSlotsConfig, toolbarConsumerConfig) => {
  const configuredToolbar = toolbarSlotsConfig ?? toolbarConsumerConfig;
  return { ...import_meter_bar_toolbar.defaultToolbarInternalConfig, ...configuredToolbar };
};
const buildMeterBarChartConfig = (props, consumerConfig, slots) => {
  const thresholds = combineThresholds(
    consumerConfig.thresholdsIndicators,
    slots.thresholds
  );
  const size = props.size ?? consumerConfig.size ?? import_default.defaultConfig.size;
  const min = props.min ?? consumerConfig.min ?? import_default.defaultConfig.min;
  const max = props.max ?? consumerConfig.max ?? import_default.defaultConfig.max;
  const tooltip = getTooltipConfig(slots, consumerConfig);
  const formatter = props.formatter ?? consumerConfig.formatter ?? import_constants.DEFAULT_FORMATTER;
  const unit = props.unit ?? consumerConfig.unit;
  const toolbar = getToolbarConfig(slots.toolbar, consumerConfig.toolbar);
  const colorRules = (consumerConfig.colorRules ?? []).filter(import_ColorRule.isValidColorRule).concat(slots.colorRules ?? []);
  const loading = props.loading ?? consumerConfig.loading ?? import_default.defaultConfig.loading;
  return {
    thresholds,
    thresholdsIndicators: thresholds,
    size,
    loading,
    min,
    max,
    tooltip,
    labelSlot: slots.label,
    valueSlot: slots.value,
    minSlot: slots.min,
    maxSlot: slots.max,
    formatter,
    unit,
    colorRules,
    toolbar
  };
};
const combineThresholds = (thresholdConfig, thresholdSlots) => {
  const slotsData = thresholdSlots?.data ? thresholdSlots.data : [];
  const consumerData = thresholdConfig?.data?.map(import_apply_threshold_defaults.applyThresholdDefaults) ?? [];
  return {
    data: slotsData.concat(consumerData),
    legend: thresholdSlots?.legend ?? thresholdConfig?.legend ?? import_default.defaultConfig.thresholdIndicators.legend
  };
};
