// @dynatrace/strato-components/charts/meter-bar/single-meter-bar/utils/iterate-slots.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var iterate_slots_exports = {};
__export(iterate_slots_exports, {
  iterateSlots: () => iterateSlots
});
module.exports = __toCommonJS(iterate_slots_exports);
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_apply_threshold_defaults = require("./apply-threshold-defaults.js");
var import_Intent = require("../../../../core/slots/Intent/Intent.js");
var import_is_slot = require("../../../../core/utils/is-slot.js");
var import_ColorRule = require("../../../core/slots/ColorRule.js");
var import_Toolbar = require("../../shared/slots/Toolbar.js");
var import_slot_handlers = require("../../shared/utils/slot-handlers.js");
var import_Label = require("../slots/Label.js");
var import_Legend = require("../slots/Legend.js");
var import_Max = require("../slots/Max.js");
var import_Min = require("../slots/Min.js");
var import_ThresholdIndicator = require("../slots/ThresholdIndicator.js");
var import_ThresholdLegend = require("../slots/ThresholdLegend.js");
var import_Tooltip = require("../slots/Tooltip.js");
var import_Value = require("../slots/Value.js");
function handleTooltipSlot(child, slots) {
  const { name, hidden, children } = child.props;
  slots.tooltip = { hidden: hidden ?? false, name, template: children };
}
function handleThresholdSlot(child, slots) {
  const threshold = (0, import_apply_threshold_defaults.applyThresholdDefaults)(child.props);
  if (slots.thresholds?.data) {
    slots.thresholds.data.push(threshold);
  } else {
    slots.thresholds = { data: [threshold] };
  }
}
function handleThresholdLegendSlot(slots) {
  const threshold = slots.thresholds ? { ...slots.thresholds, legend: true } : { data: [], legend: true };
  slots.thresholds = { ...threshold };
}
function handleLegendSlot(child, slots) {
  const legend = !(child.props.hidden ?? true);
  const threshold = slots.thresholds ? { ...slots.thresholds, legend } : { data: [], legend };
  slots.thresholds = { ...threshold };
}
function iterateSlots(children) {
  const slots = {
    colorRules: [],
    toolbar: { hidden: false }
  };
  let intents;
  import_react.Children.forEach(children, (child) => {
    if ((0, import_is_slot.isSlot)(child, import_Label.Label)) {
      (0, import_slot_handlers.handleLabelSlot)(child, slots);
    } else if ((0, import_is_slot.isSlot)(child, import_Value.Value)) {
      (0, import_slot_handlers.handleValueSlot)(child, slots);
    } else if ((0, import_is_slot.isSlot)(child, import_Min.Min)) {
      (0, import_slot_handlers.handleMinSlot)(child, slots);
    } else if ((0, import_is_slot.isSlot)(child, import_Max.Max)) {
      (0, import_slot_handlers.handleMaxSlot)(child, slots);
    } else if ((0, import_is_slot.isSlot)(child, import_ThresholdIndicator.ThresholdIndicator)) {
      handleThresholdSlot(child, slots);
    } else if ((0, import_is_slot.isSlot)(child, import_ThresholdLegend.ThresholdLegend)) {
      handleThresholdLegendSlot(slots);
    } else if ((0, import_is_slot.isSlot)(child, import_Legend.Legend)) {
      handleLegendSlot(child, slots);
    } else if ((0, import_is_slot.isSlot)(child, import_ColorRule.ColorRule)) {
      (0, import_slot_handlers.handleColorRuleSlot)(child, slots);
    } else if ((0, import_is_slot.isSlot)(child, import_Tooltip.Tooltip)) {
      handleTooltipSlot(child, slots);
    } else if ((0, import_is_slot.isSlot)(child, import_Intent.Intent)) {
      intents = (0, import_slot_handlers.handleIntentSlot)(child, intents);
    } else if ((0, import_is_slot.isSlot)(child, import_Toolbar.Toolbar)) {
      (0, import_slot_handlers.handleToolbarSlot)(child, slots);
    }
  });
  if (!(0, import_lodash_es.isUndefined)(intents) && slots.toolbar) {
    slots.toolbar.intents = intents;
  }
  return slots;
}
