// @dynatrace/strato-components/charts/meter-bar/shared/components/meter-bar-segment/MeterBarSegment.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MeterBarSegment_exports = {};
__export(MeterBarSegment_exports, {
  MeterBarSegment: () => MeterBarSegment
});
module.exports = __toCommonJS(MeterBarSegment_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_colors = __toESM(require("@dynatrace/strato-design-tokens/colors"));
var import_default = require("./types/default.js");
var import_getTokenComputedStyle = require("../../../../core/utils/getTokenComputedStyle.js");
var import_getValidColor = require("../../../../core/utils/getValidColor.js");
var import_constants = require("../../constants.js");
var import_MeterBarLayoutSizeContext = require("../../context/MeterBarLayoutSizeContext.js");
var import_element_size_css = require("../../styles/element-size.sty.js");
var import_get_dash_array = require("../../utils/get-dash-array.js");
const MeterBarSegment = (props) => {
  const { color, width, highlighted = true, start = 0 } = props;
  const validColor = (0, import_getTokenComputedStyle.getTokenComputedStyle)((0, import_getValidColor.getValidColor)(color));
  const { size, segmentHeight } = (0, import_MeterBarLayoutSizeContext.useMeterBarLayoutSize)();
  const dashArray = (0, import_get_dash_array.getDashArray)(width, segmentHeight);
  const isSegmentHeight = size === import_constants.METER_BAR_SIZE.Auto;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "rect",
    {
      className: (0, import_clsx.default)({ [(0, import_element_size_css.elementSize)({ size })]: !isSegmentHeight }),
      x: start,
      y: "0",
      height: isSegmentHeight ? segmentHeight : void 0,
      width,
      fill: validColor,
      opacity: highlighted ? import_default.HIGHLIGHTED_SEGMENT_VALUE : import_default.INACTIVE_SEGMENT_VALUE,
      "data-testid": "meterbar-segment",
      stroke: import_colors.default.Background.Base.Default,
      strokeWidth: import_constants.DEFAULT_GAP_WIDTH,
      strokeDasharray: dashArray
    }
  );
};
