// @dynatrace/strato-components/tables/DataTable/features/VisualVariants/VisualVariants.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var VisualVariants_exports = {};
__export(VisualVariants_exports, {
  VisualVariantConfig: () => VisualVariantConfig,
  useVisualVariant: () => useVisualVariant
});
module.exports = __toCommonJS(VisualVariants_exports);
var import_is_node_environment = require("../../../../core/utils/is-node-environment.js");
var import_useDataTableConfig = require("../../../hooks/useDataTableConfig.js");
var import_constants = require("../../constants.js");
var import_src = require("../../tanstack-table-core-override/src/index.js");
function isVerticalAlignmentExtendedConfig(verticalAlignment) {
  if (verticalAlignment !== void 0 && typeof verticalAlignment !== "string") {
    return true;
  }
  return false;
}
function useVisualVariant(props, options) {
  const configVariant = (0, import_useDataTableConfig.useDataTableConfigElement)("variant");
  options.state = {
    ...options.state,
    visualVariantConfig: {
      rowSeparation: props.variant?.rowSeparation ?? configVariant?.rowSeparation ?? import_constants.DATA_TABLE_VISUAL_VARIANT_CONFIG_DEFAULT.rowSeparation,
      verticalDividers: props.variant?.verticalDividers ?? configVariant?.verticalDividers ?? import_constants.DATA_TABLE_VISUAL_VARIANT_CONFIG_DEFAULT.verticalDividers,
      contained: props.variant?.contained ?? configVariant?.contained ?? import_constants.DATA_TABLE_VISUAL_VARIANT_CONFIG_DEFAULT.contained,
      rowDensity: props.variant?.rowDensity ?? configVariant?.rowDensity ?? import_constants.DATA_TABLE_VISUAL_VARIANT_CONFIG_DEFAULT.rowDensity,
      fontStyle: props.variant?.fontStyle ?? configVariant?.fontStyle ?? import_constants.DATA_TABLE_VISUAL_VARIANT_CONFIG_DEFAULT.fontStyle,
      headers: props.variant?.headers ?? configVariant?.headers ?? import_constants.DATA_TABLE_VISUAL_VARIANT_CONFIG_DEFAULT.headers,
      verticalAlignment: props.variant?.verticalAlignment ?? configVariant?.verticalAlignment ?? import_constants.DATA_TABLE_VISUAL_VARIANT_CONFIG_DEFAULT.verticalAlignment
    }
  };
}
const VisualVariantConfig = {
  getInitialState(state) {
    return {
      visualVariantConfig: {
        rowSeparation: import_constants.DATA_TABLE_VISUAL_VARIANT_CONFIG_DEFAULT.rowSeparation,
        verticalDividers: import_constants.DATA_TABLE_VISUAL_VARIANT_CONFIG_DEFAULT.verticalDividers,
        contained: import_constants.DATA_TABLE_VISUAL_VARIANT_CONFIG_DEFAULT.contained,
        rowDensity: import_constants.DATA_TABLE_VISUAL_VARIANT_CONFIG_DEFAULT.rowDensity,
        fontStyle: import_constants.DATA_TABLE_VISUAL_VARIANT_CONFIG_DEFAULT.fontStyle,
        headers: import_constants.DATA_TABLE_VISUAL_VARIANT_CONFIG_DEFAULT.headers,
        verticalAlignment: import_constants.DATA_TABLE_VISUAL_VARIANT_CONFIG_DEFAULT.verticalAlignment
      },
      ...state
    };
  },
  createTable(table) {
    table.getVisualVariantOptions = () => {
      return table.getState().visualVariantConfig;
    };
    table.getBodyVerticalAlignment = (0, import_src.memo)(
      () => [table.getVisualVariantOptions().verticalAlignment],
      (verticalAlignment) => {
        if (isVerticalAlignmentExtendedConfig(verticalAlignment)) {
          return verticalAlignment["body"];
        }
        return verticalAlignment;
      },
      { key: (0, import_is_node_environment.isDev)() && "getBodyVerticalAlignment" }
    );
    table.getHeaderVerticalAlignment = (0, import_src.memo)(
      () => [table.getVisualVariantOptions().verticalAlignment],
      (verticalAlignment) => {
        if (isVerticalAlignmentExtendedConfig(verticalAlignment)) {
          return verticalAlignment["header"];
        }
        return verticalAlignment;
      },
      { key: (0, import_is_node_environment.isDev)() && "getHeaderVerticalAlignment" }
    );
  }
};
