// @dynatrace/strato-components/overlays/tooltip/Tooltip.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Tooltip_exports = {};
__export(Tooltip_exports, {
  Tooltip: () => Tooltip
});
module.exports = __toCommonJS(Tooltip_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_framer_motion = require("framer-motion");
var import_react = require("react");
var import_TooltipElement = require("./TooltipElement.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useId = require("../../core/hooks/useId.js");
var import_useModality = require("../../core/hooks/useModality.js");
var import_usePoliteContainer = require("../hooks/usePoliteContainer.js");
var import_useTooltipState = require("../hooks/useTooltipState.js");
const COMPONENT_NAME = "Tooltip";
const Tooltip = /* @__PURE__ */ (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const {
      children,
      text,
      placement = "top",
      open,
      defaultOpen,
      onOpenChange,
      delay = "default",
      disabled,
      id: propId,
      fallbackPlacements,
      style: consumerStyle,
      className: consumerClassName,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      // @ts-expect-error - We want to allow overriding data-dt-component as we use the component internally in other Strato components, but it should not appear in the public API.
      "data-dt-component": dataDtComponent = COMPONENT_NAME,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
    const {
      "data-dt-name": dataDtName = "Trigger",
      ...remainingBehavioralTrackingProps
    } = behavioralTrackingProps;
    const generatedId = (0, import_useId.useId)();
    const id = propId ?? generatedId;
    const shouldReduceMotion = (0, import_framer_motion.useReducedMotion)();
    const {
      isOpen,
      open: openTooltip,
      close: closeTooltip
    } = (0, import_useTooltipState.useTooltipState)({
      open,
      defaultOpen,
      onOpenChange,
      delay: shouldReduceMotion || delay === "none" ? 0 : void 0
    });
    const tooltipWrapperRef = (0, import_react.useRef)(null);
    const [tooltipReferenceElement, setTooltipReferenceElement] = (0, import_react.useState)(void 0);
    const mutationObserverRef = (0, import_react.useRef)(
      new MutationObserver((entries) => {
        for (const entry of entries) {
          const firstNode = entry.addedNodes.item(0);
          if (entry.type === "childList" && firstNode !== null && firstNode instanceof Element && firstNode !== tooltipReferenceElement) {
            setTooltipReferenceElement(firstNode);
          }
        }
      })
    );
    (0, import_react.useEffect)(
      () => {
        if (tooltipWrapperRef.current) {
          setTooltipReferenceElement(tooltipWrapperRef.current?.children?.[0]);
          mutationObserverRef.current.observe(tooltipWrapperRef.current, {
            childList: true,
            subtree: false
            // only observe changes in the direct children of the wrapper, not deeper in the DOM
          });
        }
        return () => closeTooltip();
      },
      // eslint-disable-next-line react-hooks/exhaustive-deps
      []
    );
    (0, import_react.useEffect)(() => {
      if (disabled) {
        closeTooltip();
      }
    }, [closeTooltip, disabled]);
    const isTooltipHovered = (0, import_react.useRef)(false);
    const handleKeyDown = (0, import_react.useCallback)(
      (e) => e.code === "Escape" && closeTooltip(),
      [closeTooltip]
    );
    (0, import_react.useEffect)(() => {
      document.addEventListener("keydown", handleKeyDown);
      return () => {
        document.removeEventListener("keydown", handleKeyDown);
      };
    }, [handleKeyDown]);
    const PoliteContainer = (0, import_usePoliteContainer.usePoliteContainer)();
    const modality = (0, import_useModality.useModality)();
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        "span",
        {
          style: {
            // Safari has currently an outstanding issue that results in content
            // being inaccessible, when rendered within a display: contents.
            display: "contents"
          },
          ref: tooltipWrapperRef,
          onMouseEnter: () => !disabled && openTooltip(),
          onMouseLeave: () => (
            // we need to wrap this in a requestAnimationFrame to be able to first set the isTooltipHovered state correctly in case the TooltipElement (including the gap) is hovered
            // if we don't, the tooltip is closed before the onMouseOver event of the TooltipElement is triggered (to set the right state), as the onMouseLeave of the tooltip trigger would be queued up afterwards
            requestAnimationFrame(() => {
              if (!disabled && !isTooltipHovered.current) {
                closeTooltip();
              }
            })
          ),
          onFocus: () => !disabled && modality !== "pointer" && modality !== "touch" && openTooltip(true),
          onBlur: () => !disabled && !isTooltipHovered.current && closeTooltip(),
          "data-dt-name": dataDtName,
          ...remainingBehavioralTrackingProps,
          ...restProps,
          "data-dt-component": dataDtComponent,
          children
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(PoliteContainer, { children: isOpen && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_TooltipElement.TooltipElement,
        {
          ref: forwardedRef,
          triggerElement: tooltipReferenceElement,
          placement,
          fallbackPlacements,
          id,
          "data-testid": dataTestId,
          "data-dtrum-mask": dataDtrumMask,
          "data-dtrum-allow": dataDtrumAllow,
          className: consumerClassName,
          style: consumerStyle,
          onMouseLeave: (event) => !disabled && // only close the tooltip if the mouse isn't over the trigger element
          event.relatedTarget !== tooltipReferenceElement && closeTooltip(),
          onTooltipHovered: (isHovered) => isTooltipHovered.current = isHovered,
          ...behavioralTrackingProps,
          "data-dt-component": dataDtComponent,
          children: text
        }
      ) })
    ] });
  }
);
Tooltip.displayName = COMPONENT_NAME;
