// @dynatrace/strato-components/overlays/tooltip/TooltipElement.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TooltipElement_exports = {};
__export(TooltipElement_exports, {
  TooltipElement: () => TooltipElement
});
module.exports = __toCommonJS(TooltipElement_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_core = require("@popperjs/core");
var import_dynamic = require("@vanilla-extract/dynamic");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_popper = require("react-popper");
var import_tooltip_css = require("./tooltip.sty.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_useOverlayStacking = require("../../core/hooks/useOverlayStacking.js");
var import_Text = require("../../typography/text/Text.js");
var import_popper_modifiers = require("../utils/popper-modifiers.js");
const TOOLTIP_OFFSET = 4;
const TooltipElement = /* @__PURE__ */ (0, import_react.forwardRef)((props, forwardedRef) => {
  const {
    children,
    placement,
    triggerElement,
    id,
    fallbackPlacements: fallbackPlacementsProp,
    onMouseLeave,
    onTooltipHovered,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    "data-dt-component": dataDtComponent,
    ...remainingProps
  } = props;
  const [behavioralTrackingProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const {
    "data-dt-name": dataDtName = "Overlay",
    ...remainingBehavioralTrackingProps
  } = behavioralTrackingProps;
  const [tooltipRef, setTooltipRef] = (0, import_react.useState)(null);
  const mergedTooltipRef = (0, import_useMergeRefs.useMergeRefs)([forwardedRef, setTooltipRef]);
  const fallbackPlacements = (0, import_react.useMemo)(() => {
    if (fallbackPlacementsProp) {
      return fallbackPlacementsProp;
    }
    switch (placement) {
      case "top":
        return ["bottom", "left", "right"];
      case "bottom":
        return ["top", "left", "right"];
      case "left":
        return ["right", "top", "bottom"];
      case "right":
        return ["left", "top", "bottom"];
    }
  }, [placement, fallbackPlacementsProp]);
  const { styles, attributes } = (0, import_react_popper.usePopper)(triggerElement, tooltipRef, {
    placement,
    modifiers: [
      {
        name: "offset",
        options: {
          offset: [0, TOOLTIP_OFFSET]
        }
      },
      {
        name: "flip",
        options: {
          fallbackPlacements
        }
      },
      import_popper_modifiers.observeOverlaySize,
      import_popper_modifiers.observeAnimationDone,
      import_core.eventListeners
    ]
  });
  const stackingContext = (0, import_useOverlayStacking.useOverlayStacking)(void 0, "overlay");
  const tooltipHovered = (0, import_react.useRef)(false);
  const handleMouseLeaveTooltip = (0, import_react.useCallback)(
    (e) => {
      tooltipHovered.current = false;
      onMouseLeave?.(e);
      onTooltipHovered?.(tooltipHovered.current);
    },
    [onMouseLeave, onTooltipHovered]
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      ref: mergedTooltipRef,
      style: {
        ...styles.popper,
        zIndex: stackingContext.style.zIndex,
        ...consumerStyle,
        ...(0, import_dynamic.assignInlineVars)({ [import_tooltip_css.tooltipOffsetVar]: `${TOOLTIP_OFFSET}px` })
      },
      className: (0, import_clsx.default)(import_tooltip_css.tooltipElementCSS, consumerClassName),
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      role: "tooltip",
      onMouseEnter: () => {
        tooltipHovered.current = true;
        onTooltipHovered?.(tooltipHovered.current);
      },
      onMouseLeave: handleMouseLeaveTooltip,
      ...attributes.popper,
      ...remainingBehavioralTrackingProps,
      "data-dt-name": dataDtName,
      "data-dt-component": dataDtComponent,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: import_tooltip_css.tooltipAnimationWrapperCSS, children: typeof children === "string" ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { textStyle: "small", as: "p", children }) : children })
    }
  );
});
TooltipElement.displayName = "TooltipElement";
