// @dynatrace/strato-components/esm/charts/meter-bar/single-meter-bar/providers/MeterBarChartProviders.js
import { jsx } from "react/jsx-runtime";
import { useMemo } from "react";
import { OverlayProvider } from "../../../core/components/overlay-tooltip/contexts/OverlayContext.js";
import { TooltipProvider } from "../../../core/components/overlay-tooltip/contexts/TooltipContext.js";
import { ChartSeriesActionContext } from "../../../core/components/series-action/ChartSeriesActionContext.js";
import { ColorRuleContext } from "../../../core/context/ColorRuleContext.js";
import { FormatterContext } from "../../shared/context/FormatterContext.js";
import { MeterBarChartLoadingContext } from "../../shared/context/MeterBarChartLoadingContext.js";
import { MeterBarToolbarConfigContext } from "../../shared/context/MeterBarToolbarConfigContext.js";
import {
  MeterBarChartInternalConfigContext
} from "../context/MeterBarChartInternalConfigContext.js";
const MeterBarChartProviders = ({
  config,
  seriesActions,
  children
}) => {
  const contextSeriesActionsValue = useMemo(
    () => ({
      chartSeriesActions: seriesActions
    }),
    [seriesActions]
  );
  const internalConfigValue = useMemo(
    () => ({
      min: config.min,
      max: config.max,
      tooltip: config.tooltip,
      label: config.labelSlot,
      thresholds: config.thresholds,
      thresholdsIndicators: config.thresholdsIndicators
    }),
    [config]
  );
  return /* @__PURE__ */ jsx(ColorRuleContext.Provider, { value: config.colorRules, children: /* @__PURE__ */ jsx(
    FormatterContext.Provider,
    {
      value: { formatter: config.formatter, unit: config.unit },
      children: /* @__PURE__ */ jsx(MeterBarChartLoadingContext.Provider, { value: config.loading, children: /* @__PURE__ */ jsx(ChartSeriesActionContext.Provider, { value: contextSeriesActionsValue, children: /* @__PURE__ */ jsx(
        MeterBarChartInternalConfigContext.Provider,
        {
          value: internalConfigValue,
          children: /* @__PURE__ */ jsx(MeterBarToolbarConfigContext.Provider, { value: config.toolbar, children: /* @__PURE__ */ jsx(OverlayProvider, { children: /* @__PURE__ */ jsx(TooltipProvider, { children }) }) })
        }
      ) }) })
    }
  ) });
};
export {
  MeterBarChartProviders
};
