// @dynatrace/strato-components/esm/charts/single-value-grid/defaults.js
const DISPLAY_NAME = "SingleValueGrid";
const sparklineDefaultProps = {
  showTicks: false,
  showContextValues: false,
  variant: "line",
  curve: "linear",
  gapPolicy: "connect",
  className: "",
  style: {},
  "data-testid": "",
  "data-dtrum-mask": false,
  "data-dtrum-allow": false,
  "data-dt-name": "",
  "data-dt-content": "",
  "data-dt-children-name": "",
  "data-dt-features": "",
  "data-dt-properties": "",
  "data-dt-mouse-over": "",
  "data-dt-value-content": false,
  "data-dt-children-value-content": false,
  "data-dt-ancestor": ""
};
const trendDefaultProps = {
  showIcon: true,
  inverseTrend: "none"
};
const gridDefaultProps = {
  alignment: "start",
  applyThresholdBackground: false,
  className: "",
  style: {},
  "data-dtrum-mask": false,
  "data-dtrum-allow": false,
  "data-testid": "",
  loading: false,
  "data-dt-name": "",
  "data-dt-content": "",
  "data-dt-children-name": "",
  "data-dt-features": "",
  "data-dt-properties": "",
  "data-dt-mouse-over": "",
  "data-dt-value-content": false,
  "data-dt-children-value-content": false,
  "data-dt-ancestor": ""
};
const SINGLE_VALUE_GRID_ARIA_LABEL = "Single value data visualization grid";
export {
  DISPLAY_NAME,
  SINGLE_VALUE_GRID_ARIA_LABEL,
  gridDefaultProps,
  sparklineDefaultProps,
  trendDefaultProps
};
