// @dynatrace/strato-components/esm/buttons/button/Button.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import {
  Children,
  forwardRef,
  useMemo
} from "react";
import { useIntl } from "react-intl";
import { isElement } from "react-is";
import * as styles from "./Button.sty.js";
import { Label } from "./Label.js";
import { Prefix } from "./Prefix.js";
import { Suffix } from "./Suffix.js";
import { ProgressCircle } from "../../content/progress/ProgressCircle.js";
import { useAriaLabelingProps } from "../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { focusRingStatic } from "../../core/styles/focusRingStatic.sty.js";
import { mergeProps } from "../../core/utils/merge-props.js";
import { parseBoolean } from "../../core/utils/parse-boolean.js";
import { Flex } from "../../layouts/flex/Flex.js";
import { Grid } from "../../layouts/grid/Grid.js";
import { fieldTokens } from "../../styles/field.sty.js";
const COMPONENT_NAME = "Button";
function getButtonSlots(children) {
  const slots = {};
  Children.forEach(children, (child) => {
    if (isElement(child)) {
      switch (child.type) {
        case Suffix:
          slots.suffix = child;
          break;
        case Prefix:
          slots.prefix = child;
          break;
        case Label:
          slots.label = child;
          break;
        default:
          slots.freeform ??= [];
          slots.freeform.push(child);
          break;
      }
    } else {
      if (child !== null && child !== void 0) {
        slots.freeform ??= [];
        slots.freeform.push(child);
      }
    }
  });
  return slots;
}
const _Button = /* @__PURE__ */ forwardRef(
  (props, ref) => {
    const {
      children,
      variant = "default",
      disabled,
      "aria-disabled": ariaDisabledProp,
      required,
      id: propId,
      type = "button",
      onClick,
      ariaLabel,
      width = "content",
      color = "neutral",
      textAlign = "center",
      size = "default",
      loading = false,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      as,
      ...remainingProps
    } = props;
    const slots = getButtonSlots(children);
    const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
    const ariaDisabled = parseBoolean(ariaDisabledProp);
    const nativeDisabledOrLoading = useMemo(
      () => loading ? true : disabled,
      [disabled, loading]
    );
    const disabledOrLoading = useMemo(
      () => nativeDisabledOrLoading || ariaDisabled,
      [ariaDisabled, nativeDisabledOrLoading]
    );
    const ariaLabelingProps = useAriaLabelingProps(restProps);
    const handleButtonClick = (event) => {
      if (nativeDisabledOrLoading) {
        event.preventDefault();
        event.stopPropagation();
        return;
      }
      onClick?.(event);
    };
    const buttonProps = {
      ref,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      required,
      disabled: nativeDisabledOrLoading && !ariaDisabled,
      id: propId,
      type,
      ...ariaLabelingProps,
      "aria-disabled": ariaDisabledProp,
      onClick: handleButtonClick
    };
    const disabledPropagationProps = {
      onPointerDown(evt) {
        if (nativeDisabledOrLoading) {
          evt.preventDefault();
          evt.stopPropagation();
        }
      },
      onPointerUp(evt) {
        if (nativeDisabledOrLoading) {
          evt.preventDefault();
          evt.stopPropagation();
        }
      }
    };
    const hasChildren = children || children === 0;
    const buttonMergedProps = {
      ...mergeProps(
        buttonProps,
        disabledPropagationProps,
        // Need to cast this one as the inference of mergeProps would generate
        // a not allowed overlap. The restProps type is too complex due to the
        // polymorphic inheritance.
        restProps,
        behavioralTrackingProps
      )
    };
    const intl = useIntl();
    const Component = as || "button";
    if (!hasChildren) {
      return /* @__PURE__ */ jsx(
        Component,
        {
          "data-dt-component": COMPONENT_NAME,
          ...buttonMergedProps,
          className: clsx(
            consumerClassName,
            "strato-inputgroup",
            fieldTokens({ variant, color, disabled: disabledOrLoading }),
            styles.button({
              disabled: disabledOrLoading,
              textAlign,
              size
            }),
            focusRingStatic({
              when: "focus-visible",
              color,
              disabled: nativeDisabledOrLoading
            }),
            {
              [`${styles.widths.content}`]: width === "content",
              [`${styles.widths.full}`]: width === "full"
            }
          ),
          style: {
            width: width !== "content" && width !== "full" ? width : void 0
          }
        }
      );
    }
    return /* @__PURE__ */ jsxs(
      Component,
      {
        "data-dt-component": COMPONENT_NAME,
        className: clsx(
          consumerClassName,
          "strato-inputgroup",
          fieldTokens({
            variant,
            color,
            disabled: disabledOrLoading || ariaDisabled
          }),
          styles.button({
            disabled: disabledOrLoading || ariaDisabled,
            textAlign,
            size
          }),
          {
            [`${styles.widths.content}`]: width === "content",
            [`${styles.widths.full}`]: width === "full"
          },
          focusRingStatic({
            when: "focus-visible",
            color,
            disabled: nativeDisabledOrLoading
          })
        ),
        style: {
          width: width !== "content" && width !== "full" ? width : void 0,
          ...consumerStyle
        },
        ...buttonMergedProps,
        children: [
          loading && /* @__PURE__ */ jsx("div", { className: styles.loading, children: /* @__PURE__ */ jsx(
            ProgressCircle,
            {
              size: "small",
              color,
              "aria-label": intl.formatMessage({
                defaultMessage: "Loading",
                id: "cycz+bYG+VH7Of4e",
                description: "Label for the loading state."
              })
            }
          ) }),
          /* @__PURE__ */ jsxs(
            Grid,
            {
              gap: size === "default" ? 4 : 2,
              grid: "auto / auto-flow auto",
              width: "100%",
              maxWidth: "100%",
              alignItems: "center",
              justifyContent: textAlign === "start" ? "space-between" : "center",
              style: {
                visibility: loading ? "hidden" : void 0
              },
              children: [
                textAlign === "start" ? /* @__PURE__ */ jsxs(Flex, { minWidth: 0, gap: size === "default" ? 4 : 2, children: [
                  slots.prefix,
                  slots.label,
                  slots.freeform && /* @__PURE__ */ jsx(Button.Label, { children: slots.freeform })
                ] }) : /* @__PURE__ */ jsxs(Fragment, { children: [
                  slots.prefix,
                  slots.label,
                  slots.freeform && /* @__PURE__ */ jsx(Button.Label, { children: slots.freeform })
                ] }),
                slots.suffix
              ]
            }
          )
        ]
      }
    );
  }
);
_Button.displayName = COMPONENT_NAME;
const Button = Object.assign(_Button, {
  Suffix,
  Prefix,
  Label
});
export {
  Button,
  getButtonSlots as default
};
