// @dynatrace/strato-components/filters/segment-selector/parts/noAccessRights/NoAccessWithoutSelection.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var NoAccessWithoutSelection_exports = {};
__export(NoAccessWithoutSelection_exports, {
  NoAccessWithoutSelection: () => NoAccessWithoutSelection
});
module.exports = __toCommonJS(NoAccessWithoutSelection_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_Button = require("../../../../buttons/button/Button.js");
var import_external_links = require("../../../../core/external-links.js");
var import_Flex = require("../../../../layouts/flex/Flex.js");
var import_Text = require("../../../../typography/text/Text.js");
var import_segmentMessages = require("../../segmentMessages.js");
var styles = __toESM(require("../Overlay.sty.js"));
var import_CopyPermissionsButton = require("./CopyPermissionsButton.js");
const NoAccessWithoutSelection = () => {
  const intl = (0, import_react_intl.useIntl)();
  const buttonRef = (0, import_react.useRef)(null);
  (0, import_react.useEffect)(() => {
    buttonRef.current?.focus();
  }, []);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_Flex.Flex,
    {
      flexDirection: "column",
      alignItems: "center",
      padding: 16,
      width: "100vw",
      className: styles.overlayDefaultWidth,
      style: { maxWidth: "320px", textAlign: "center" },
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Text.Text, { textStyle: "base-emphasized", children: intl.formatMessage(import_segmentMessages.SEGMENTSELECTOR_MESSAGES.noAccessNoSegments) }),
        intl.formatMessage(import_segmentMessages.SEGMENTSELECTOR_MESSAGES.copyDetailsHint),
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_Flex.Flex, { children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_CopyPermissionsButton.CopyPermissionsButton, { variant: "emphasized", ref: buttonRef }),
          /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
            import_Button.Button,
            {
              as: "a",
              href: import_external_links.SEGMENTS_DOCUMENTATION_URL,
              target: "_blank",
              rel: "noopener noreferrer",
              children: [
                intl.formatMessage(import_segmentMessages.SEGMENTSELECTOR_MESSAGES.docsLinkTextShort),
                /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Suffix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ExternalLinkIcon, {}) })
              ]
            }
          )
        ] })
      ]
    }
  );
};
