// @dynatrace/strato-components/notifications/index.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var notifications_exports = {};
__export(notifications_exports, {
  NotificationSettings: () => import_NotificationSettings.NotificationSettings,
  ToastContainer: () => import_ToastContainer.ToastContainer,
  dismissAllToasts: () => import_Toast.dismissAllToasts,
  dismissToast: () => import_Toast.dismissToast,
  showToast: () => import_Toast.showToast,
  useNotificationSettingsMenuItem: () => import_useNotificationSettingsMenuItem.useNotificationSettingsMenuItem
});
module.exports = __toCommonJS(notifications_exports);
var import_Toast = require("./toast/Toast.js");
var import_ToastContainer = require("./toast/ToastContainer.js");
var import_NotificationSettings = require("./notification-settings/NotificationSettings.js");
var import_useNotificationSettingsMenuItem = require("./notification-settings/hooks/useNotificationSettingsMenuItem.js");
