// @dynatrace/strato-components/forms/checkbox/Checkbox.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Checkbox_exports = {};
__export(Checkbox_exports, {
  Checkbox: () => Checkbox
});
module.exports = __toCommonJS(Checkbox_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_Checkbox_css = require("./Checkbox.sty.js");
var import_VisuallyHidden = require("../../core/components/visually-hidden/VisuallyHidden.js");
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useControllableState = require("../../core/hooks/useControllableState.js");
var import_useFormControlState = require("../../core/hooks/useFormControlState.js");
var import_useSafeForwardProps = require("../../core/hooks/useSafeForwardProps.js");
var import_focusRingStatic_css = require("../../core/styles/focusRingStatic.sty.js");
var import_merge_props = require("../../core/utils/merge-props.js");
var import_Flex = require("../../layouts/flex/Flex.js");
var import_field_css = require("../../styles/field.sty.js");
var import_textStyle_css = require("../../styles/textStyle.sty.js");
var import_FormFieldMessageBaseItem = require("../forms-core/components/form-field-messages/FormFieldMessageBaseItem.js");
var import_useFormFieldMessagesContext = require("../forms-core/hooks/useFormFieldMessagesContext.js");
var import_useFormFieldValidation = require("../forms-core/hooks/useFormFieldValidation.js");
var import_useSyncedFormControlProps = require("../forms-core/hooks/useSyncedFormControlProps.js");
var import_formfield_messages_state_reducer = require("../forms-core/state/formfield-messages-state-reducer.js");
var import_useInputHint = require("../hooks/useInputHint.js");
const COMPONENT_NAME = "Checkbox";
const Checkbox = (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const {
      disabled: propDisabled,
      required: propRequired,
      readOnly: propReadOnly,
      id: propId,
      defaultValue,
      value,
      name,
      formValue,
      onChange,
      children,
      controlState,
      className: consumerClassName,
      style: consumerStyle,
      "aria-disabled": ariaDisabledProp,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
    const ariaLabelingProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(restProps);
    const forwardProps = (0, import_useSafeForwardProps.useSafeForwardProps)(restProps, ariaLabelingProps);
    const { state: messagesContextState, dispatch: messagesContextDispatch } = (0, import_useFormFieldMessagesContext.useFormFieldMessagesContext)();
    const {
      disabled,
      required,
      id,
      ariaDisabled,
      readOnly: _readOnly
    } = (0, import_useSyncedFormControlProps.useSyncedFormControlProps)({
      propDisabled,
      propRequired,
      propId,
      propAriaDisabled: ariaDisabledProp,
      propReadOnly
    });
    const rootRef = (0, import_react.useRef)(null);
    const inputRef = (0, import_react.useRef)(null);
    const {
      onChange: validationOnChange,
      onBlur: validationOnBlur,
      onInvalid: validationOnInvalid,
      ariaProps
    } = (0, import_useFormFieldValidation.useFormFieldValidation)(inputRef, { required }, ariaLabelingProps);
    (0, import_react.useImperativeHandle)(forwardedRef, () => {
      return {
        element: rootRef.current,
        inputRef: inputRef.current,
        focus() {
          inputRef.current?.focus();
        },
        validate() {
          if (messagesContextState) {
            inputRef.current?.checkValidity();
            messagesContextDispatch?.({
              type: import_formfield_messages_state_reducer.FormFieldMessagesAction.SETISTOUCHED,
              payload: { isTouched: true }
            });
          }
        }
      };
    });
    const [state, setState] = (0, import_useControllableState.useControllableState)({
      defaultValue,
      value
    });
    const disabledOrAriaDisabled = disabled || ariaDisabled || false;
    const readOnly = !!_readOnly;
    const [hasError, hint] = (0, import_useFormControlState.useFormControlState)(controlState);
    const { inputProps, hintProps } = (0, import_useInputHint.useInputHint)({
      hint,
      hasError
    });
    const showError = hasError || messagesContextState.showError;
    let focusRingColor = "neutral";
    if (showError) {
      focusRingColor = "critical";
    } else if (!(disabled || readOnly) && state) {
      focusRingColor = "primary";
    }
    const handleOnChange = (e) => {
      if (readOnly) {
        return;
      }
      const newValue = e.target.indeterminate ? true : e.target.checked;
      setState(newValue);
      onChange?.(newValue, e);
      validationOnChange();
      if (!messagesContextState.isTouched) {
        messagesContextDispatch?.({
          type: import_formfield_messages_state_reducer.FormFieldMessagesAction.SETISTOUCHED,
          payload: { isTouched: true }
        });
      }
    };
    (0, import_react.useEffect)(() => {
      if (inputRef.current) {
        inputRef.current.indeterminate = value === "indeterminate";
      }
    }, [value]);
    let fieldColorVariant = "default";
    if (state === true || state === "indeterminate") {
      fieldColorVariant = "accent";
    } else if (showError || disabledOrAriaDisabled) {
      fieldColorVariant = "emphasized";
    }
    let fieldColorColor = "primary";
    if (showError) {
      fieldColorColor = "critical";
    } else if (state === false) {
      fieldColorColor = "neutral";
    }
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      import_Flex.Flex,
      {
        ref: rootRef,
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        flexDirection: "column",
        gap: 4,
        className: consumerClassName,
        style: consumerStyle,
        ...(0, import_merge_props.mergeProps)(forwardProps, behavioralTrackingProps),
        "data-dt-component": COMPONENT_NAME,
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
            "label",
            {
              htmlFor: id,
              className: (0, import_clsx.default)(
                (0, import_Checkbox_css.checkboxLabelCSS)({
                  required: typeof children === "string" && required,
                  disabled: disabledOrAriaDisabled,
                  readOnly
                }),
                (0, import_textStyle_css.textStyle)({ fontStyle: "text", textStyle: "base" })
              ),
              children: [
                /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
                  "div",
                  {
                    className: (0, import_clsx.default)(
                      (0, import_Checkbox_css.checkboxKnobCSS)({
                        disabled: disabledOrAriaDisabled,
                        hasChildren: !!children,
                        hasError: showError,
                        checked: state === true || state === "indeterminate",
                        readOnly
                      }),
                      (0, import_field_css.fieldTokens)({
                        variant: fieldColorVariant,
                        disabled: disabledOrAriaDisabled,
                        color: fieldColorColor
                      }),
                      (0, import_focusRingStatic_css.focusRingStatic)({
                        when: "focus-within-visible",
                        color: focusRingColor,
                        disabled
                      })
                    ),
                    children: [
                      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_VisuallyHidden.VisuallyHidden, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                        "input",
                        {
                          id,
                          ref: inputRef,
                          type: "checkbox",
                          "aria-checked": state === "indeterminate" ? "mixed" : state,
                          "aria-required": required,
                          "aria-disabled": ariaDisabled,
                          readOnly,
                          name,
                          value: formValue,
                          disabled,
                          required,
                          checked: state === true,
                          onChange: handleOnChange,
                          onBlur: (event) => {
                            validationOnBlur(event);
                          },
                          onInvalid: validationOnInvalid,
                          ...controlState ? inputProps : ariaProps
                        }
                      ) }),
                      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: (0, import_Checkbox_css.checkboxIconCSS)({ readOnly, hasError: showError }), children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(CheckboxIconSvg, { state }) })
                    ]
                  }
                ),
                children
              ]
            }
          ),
          hint && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_FormFieldMessageBaseItem.InternalBaseItem,
            {
              variant: showError ? "error" : "hint",
              disabled: disabledOrAriaDisabled,
              ...hintProps,
              children: hint
            }
          )
        ]
      }
    );
  }
);
Checkbox.displayName = COMPONENT_NAME;
const CheckboxIconSvg = (0, import_react.memo)((props) => {
  switch (props.state) {
    case true:
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.CheckmarkSmallIcon, {});
    case "indeterminate":
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.MinusIcon, {});
    case false:
    default:
      return null;
  }
});
