// @dynatrace/strato-components/charts/histogram/providers/HistogramDataProvider.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var HistogramDataProvider_exports = {};
__export(HistogramDataProvider_exports, {
  HistogramDataProvider: () => HistogramDataProvider,
  safeGet: () => safeGet,
  useBuildInitialMetadata: () => useBuildInitialMetadata
});
module.exports = __toCommonJS(HistogramDataProvider_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_useColorRule = require("../../core/color-rule/useColorRule.js");
var import_get_color_palette = require("../../core/utils/get-color-palette.js");
var import_HistogramData_context = require("../contexts/HistogramData.context.js");
var import_useHistogramColorPalette = require("../hooks/useHistogramColorPalette.js");
const LEFT_SCALE_INDEX = 0;
function useBuildInitialMetadata(data) {
  const allUnits = /* @__PURE__ */ new Set();
  const colorPalette = (0, import_useHistogramColorPalette.useHistogramColorPalette)();
  const seriesMap = /* @__PURE__ */ new WeakMap();
  const ruleColor = (0, import_useColorRule.useColorRule)(data, "histogram");
  for (let i = 0; i < data.length; i++) {
    const series = data[i];
    const id = crypto.randomUUID();
    const color = ruleColor.for(series) || (series.color ? { Default: series.color } : (0, import_get_color_palette.getColorPalette)(series.name, i, colorPalette));
    allUnits.add(series.unit);
    const metadata = {
      id,
      color,
      scaleIndex: LEFT_SCALE_INDEX
    };
    seriesMap.set(series, metadata);
  }
  return {
    yUnits: [...allUnits],
    seriesMap
  };
}
function safeGet(map, key) {
  const metadata = map.get(key);
  if (!metadata) {
    throw new Error(`Metadata not found for: ${JSON.stringify(key)}`);
  }
  return metadata;
}
const HistogramDataProvider = ({
  children,
  data
}) => {
  const { yUnits, seriesMap } = useBuildInitialMetadata(data);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_HistogramData_context.HistogramDataContext.Provider,
    {
      value: {
        series: data,
        seriesMap,
        yUnits
      },
      children
    }
  );
};
HistogramDataProvider["displayName"] = "HistogramDataProvider";
