// @dynatrace/strato-components/charts/histogram/providers/HistogramChartImperativeHandler.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var HistogramChartImperativeHandler_exports = {};
__export(HistogramChartImperativeHandler_exports, {
  HistogramChartImperativeHandler: () => HistogramChartImperativeHandler
});
module.exports = __toCommonJS(HistogramChartImperativeHandler_exports);
var import_react = require("react");
var import_useGridPlotSize = require("../../core/components/chart-grid/hooks/useGridPlotSize.js");
var import_types = require("../../core/components/interactions-overlay/types.js");
var import_mode_reducers = require("../../core/components/interactions-state-helpers/reducers/mode-reducers.js");
var import_useTriggerZoom = require("../../core/components/zoom/hooks/useTriggerZoom.js");
var import_chart_types = require("../../core/types/chart-types.js");
var import_useApplyZoomDelta = require("../components/overlays/hooks/useApplyZoomDelta.js");
var import_selectors = require("../hooks/store/selectors.js");
var import_state_setters = require("../hooks/store/state-setters.js");
var import_useHistogramData = require("../hooks/useHistogramData.js");
var import_useResetInteractions = require("../hooks/useResetInteractions.js");
var import_build_and_download_csv = require("../utils/build-and-download-csv.js");
var import_histogram_configuration = require("../utils/histogram-configuration.js");
const HistogramChartImperativeHandler = ({
  forwardedRef,
  chartRef,
  config,
  children
}) => {
  const { interactions } = config;
  const setState = (0, import_state_setters.useSetState)();
  const { width } = (0, import_useGridPlotSize.useGridPlotSize)();
  const applyZoom = (0, import_useApplyZoomDelta.useApplyZoomDelta)();
  const triggerZoom = (0, import_useTriggerZoom.useTriggerZoom)(width, applyZoom);
  const resetInteractions = (0, import_useResetInteractions.useResetInteractions)();
  const { series } = (0, import_useHistogramData.useHistogramData)();
  const { zoomInEnabled, zoomOutEnabled } = (0, import_selectors.useZoomState)();
  (0, import_react.useImperativeHandle)(forwardedRef, () => {
    return {
      element: chartRef.current,
      getConfig: () => JSON.stringify({
        type: import_chart_types.ChartType.Histogram,
        ...(0, import_histogram_configuration.normalizeConfig)(config)
      }),
      downloadData: () => {
        (0, import_build_and_download_csv.buildAndDownloadCsv)(series);
      },
      exploreMode: () => {
        setState((0, import_mode_reducers.reduceInteractionMode)(import_types.INTERACTION_MODE.Explore));
      },
      zoomMode: () => {
        if (interactions?.zoom) {
          setState((0, import_mode_reducers.reduceInteractionMode)(import_types.INTERACTION_MODE.Zoom));
        }
      },
      panMode: () => {
        if (interactions?.pan && zoomOutEnabled) {
          setState((0, import_mode_reducers.reduceInteractionMode)(import_types.INTERACTION_MODE.Pan));
        }
      },
      zoomIn: () => {
        if (zoomInEnabled) {
          triggerZoom("zoom-in");
        }
      },
      zoomOut: () => {
        if (zoomOutEnabled) {
          triggerZoom("zoom-out");
        }
      },
      reset: () => {
        resetInteractions();
      }
    };
  });
  return children;
};
