// @dynatrace/strato-components/charts/histogram/providers/HistogramFilteredDataProvider.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var HistogramFilteredDataProvider_exports = {};
__export(HistogramFilteredDataProvider_exports, {
  HistogramFilteredDataProvider: () => HistogramFilteredDataProvider,
  setMinMaxPerUnit: () => setMinMaxPerUnit,
  useBuildFilteredMetadata: () => useBuildFilteredMetadata
});
module.exports = __toCommonJS(HistogramFilteredDataProvider_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_HistogramDataProvider = require("./HistogramDataProvider.js");
var import_get_relative_value = require("../../core/utils/get-relative-value.js");
var import_datapoint_stacking = require("../components/displays/utils/datapoint-stacking.js");
var import_HistogramFilteredData_context = require("../contexts/HistogramFilteredData.context.js");
var import_selectors = require("../hooks/store/selectors.js");
var import_useHistogramData = require("../hooks/useHistogramData.js");
var import_useHistogramValueRepresentation = require("../hooks/useHistogramValueRepresentation.js");
var import_calculate_values = require("../utils/calculate-values.js");
const calculateTotalValuePerSeries = (totalValues, key, value) => totalValues.set(key, (totalValues.get(key) ?? 0) + value);
const getUnits = (series) => series.flatMap(
  ({ unit }) => typeof unit === "string" ? unit : unit?.map(({ group }) => group)
);
const setMinMaxPerUnit = (minMaxPerUnit, unit, binsCumulative) => {
  const currentMinMax = minMaxPerUnit.get(unit);
  if (currentMinMax) {
    minMaxPerUnit.set(unit, {
      min: Math.min(currentMinMax.min, binsCumulative),
      max: Math.max(currentMinMax.max, binsCumulative)
    });
  } else {
    minMaxPerUnit.set(unit, {
      min: Math.min(Infinity, binsCumulative),
      max: Math.max(-Infinity, binsCumulative)
    });
  }
};
const useBuildFilteredMetadata = (series, initSeriesMap) => {
  const binsMap = /* @__PURE__ */ new WeakMap();
  const filteredUnit = [];
  const valueRepresentation = (0, import_useHistogramValueRepresentation.useHistogramValueRepresentation)();
  const cumulativeStackedBins = /* @__PURE__ */ new Map();
  const minMaxPerUnit = /* @__PURE__ */ new Map();
  const minMaxXValue = { min: Infinity, max: -Infinity };
  const memoizedMaxSeriesValue = /* @__PURE__ */ new Map();
  const isRelative = valueRepresentation === "relative";
  const totalValues = /* @__PURE__ */ new Map();
  const units = isRelative ? new Set(getUnits(series)) : /* @__PURE__ */ new Set();
  if (isRelative) {
    series.forEach(
      ({ bins }) => bins.forEach(
        ({ from, to, value }) => calculateTotalValuePerSeries(totalValues, `${from}-${to}`, value)
      )
    );
  }
  for (let i = 0; i < series.length; i++) {
    const singleSeries = series[i];
    const hasSingleSeries = memoizedMaxSeriesValue.has(singleSeries);
    const unit = isRelative ? "%" : singleSeries.unit;
    const unitIndex = filteredUnit.findIndex((u) => u === unit);
    const isNewUnit = unitIndex < 0;
    if (isNewUnit) {
      filteredUnit.push(unit);
    }
    const { color, id } = (0, import_HistogramDataProvider.safeGet)(initSeriesMap, singleSeries);
    const scaleIndex = isNewUnit ? filteredUnit.length - 1 : unitIndex;
    let totalValue;
    if (isRelative) {
      totalValue = hasSingleSeries ? memoizedMaxSeriesValue.get(singleSeries) : (0, import_calculate_values.getMaxSeriesValue)(singleSeries);
      if (!hasSingleSeries) {
        memoizedMaxSeriesValue.set(singleSeries, totalValue);
      }
    }
    for (let j = 0; j < singleSeries.bins.length; j++) {
      const bin = singleSeries.bins[j];
      const { value, to, from } = bin;
      if (isRelative && series.length > 1 && units.size === 1) {
        totalValue = totalValues.get(`${from}-${to}`);
      }
      const newValue = isRelative ? (0, import_get_relative_value.getRelativeValue)(value, Math.abs(totalValue)) : value;
      minMaxXValue.min = Math.min(minMaxXValue.min, from, to);
      minMaxXValue.max = Math.max(minMaxXValue.max, from, to);
      const cumulativeStack = (0, import_datapoint_stacking.getCumulativeValue)(
        cumulativeStackedBins,
        bin,
        singleSeries.unit,
        id,
        newValue
      );
      const binMetadata = {
        seriesId: id,
        color,
        unit,
        value: newValue,
        from,
        to,
        seriesName: singleSeries.name,
        cumulativeValue: cumulativeStack.cumulativeValue,
        absolute: isRelative ? {
          value: bin.value,
          unit: singleSeries.unit
        } : void 0
      };
      (0, import_datapoint_stacking.setCumulativeBins)(
        cumulativeStackedBins,
        id,
        cumulativeStack.key,
        cumulativeStack.binsCumulative,
        cumulativeStack.baseCumulative
      );
      binsMap.set(bin, binMetadata);
      setMinMaxPerUnit(minMaxPerUnit, unit, cumulativeStack.binsCumulative);
    }
    const metadata = {
      id,
      color,
      scaleIndex
    };
    initSeriesMap.set(singleSeries, metadata);
  }
  return {
    binsMap,
    seriesMap: initSeriesMap,
    filteredUnits: [...filteredUnit],
    minMaxPerUnit,
    minMaxXValue
  };
};
const HistogramFilteredDataProvider = ({ children }) => {
  const { series, seriesMap: initialSeriesMap } = (0, import_useHistogramData.useHistogramData)();
  const filteredSeriesIds = (0, import_selectors.useFilteredSeriesIdsState)();
  const hasId = (series2) => filteredSeriesIds.has((0, import_HistogramDataProvider.safeGet)(initialSeriesMap, series2).id);
  const filteredHistogramSeries = filteredSeriesIds.size > 0 ? series.filter(hasId) : series;
  const { binsMap, seriesMap, filteredUnits, minMaxPerUnit, minMaxXValue } = useBuildFilteredMetadata(filteredHistogramSeries, initialSeriesMap);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_HistogramFilteredData_context.HistogramFilteredDataContext.Provider,
    {
      value: {
        series: filteredHistogramSeries,
        seriesMap,
        binsMap,
        units: filteredUnits,
        minMaxPerUnit,
        minMaxXValue
      },
      children
    }
  );
};
HistogramFilteredDataProvider["displayName"] = "HistogramFilteredDataProvider";
