// @dynatrace/strato-components/overlays/utils/popper-modifiers.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var popper_modifiers_exports = {};
__export(popper_modifiers_exports, {
  isOverlaySizeStrategyRange: () => isOverlaySizeStrategyRange,
  observeAnimationDone: () => observeAnimationDone,
  observeOverlaySize: () => observeOverlaySize,
  overlayHeightStrategyHandling: () => overlayHeightStrategyHandling,
  overlayWidthStrategyHandling: () => overlayWidthStrategyHandling
});
module.exports = __toCommonJS(popper_modifiers_exports);
var import_isBrowser = require("../../core/utils/isBrowser.js");
function isPopperEnrichedElement(value) {
  return value !== void 0 && value["__popperInstance__"] !== void 0;
}
function isOverlaySizeStrategyRange(widthStrategy) {
  return widthStrategy.min !== void 0 || widthStrategy.max !== void 0;
}
const observeOverlaySize = {
  name: "observeOverlaySize",
  enabled: true,
  phase: "main",
  options: {
    resizeObserverInstance: import_isBrowser.isBrowser && window.ResizeObserver && new ResizeObserver(async (entries) => {
      for (const entry of entries) {
        const { target } = entry;
        isPopperEnrichedElement(target) && await target["__popperInstance__"].update();
      }
    })
  },
  // needed a function even though it seems optional
  fn: () => void 0,
  effect: ({ state, options, instance }) => {
    const { popper: overlayElement, reference } = state.elements;
    const { resizeObserverInstance } = options;
    if (resizeObserverInstance instanceof ResizeObserver) {
      resizeObserverInstance?.observe(overlayElement);
      overlayElement["__popperInstance__"] = instance;
      if (reference instanceof HTMLElement) {
        resizeObserverInstance?.observe(reference);
        reference["__popperInstance__"] = instance;
      }
      return () => {
        resizeObserverInstance?.unobserve(overlayElement);
        if (isPopperEnrichedElement(overlayElement)) {
          delete overlayElement["__popperInstance__"];
        }
        if (reference instanceof HTMLElement) {
          resizeObserverInstance?.unobserve(reference);
          if (isPopperEnrichedElement(overlayElement)) {
            delete reference["__popperInstance__"];
          }
        }
      };
    }
  }
};
const observeAnimationDone = {
  name: "observeAnimationDone",
  enabled: true,
  phase: "main",
  // needed a function even though it seems optional (throws a console error otherwise)
  fn: () => void 0,
  effect: ({ state, instance }) => {
    const { popper: overlayElement } = state.elements;
    overlayElement["__popperInstance__"] = instance;
    const updatePosition = () => {
      if (isPopperEnrichedElement(overlayElement)) {
        requestAnimationFrame(async () => await instance.update());
      }
    };
    window?.addEventListener("animationDone", updatePosition, false);
    return () => {
      window?.removeEventListener("animationDone", updatePosition, false);
      if (isPopperEnrichedElement(overlayElement)) {
        delete overlayElement["__popperInstance__"];
      }
    };
  }
};
const overlayWidthStrategyHandling = {
  name: "overlayWidthStrategyHandling",
  fn: ({ state, options }) => {
    const widthStrategy = options.widthStrategy;
    if (widthStrategy === "trigger") {
      state.styles.popper.width = `${state.rects.reference.width}px`;
    } else if (isOverlaySizeStrategyRange(widthStrategy)) {
      state.styles.popper.minWidth = `${widthStrategy.min || 0}px`;
      state.styles.popper.maxWidth = widthStrategy.max !== void 0 ? `${widthStrategy.max}px` : "100%";
    }
  },
  phase: "beforeWrite",
  requires: ["computeStyles"]
};
const overlayHeightStrategyHandling = {
  name: "overlayHeightStrategyHandling",
  fn: ({ state, options }) => {
    const heightStrategy = options.heightStrategy;
    if (heightStrategy === "maxViewport") {
      let maxHeight = state.rects.reference.y;
      const spacing = 10;
      const viewportHeight = Math.max(
        document.documentElement.clientHeight || 0,
        window.innerHeight || 0
      );
      if (maxHeight < viewportHeight / 2) {
        maxHeight = viewportHeight - maxHeight - state.rects.reference.height;
      }
      state.styles.popper.maxHeight = `${maxHeight - spacing}px`;
    } else if (isOverlaySizeStrategyRange(heightStrategy)) {
      state.styles.popper.minHeight = `${heightStrategy.min || 0}px`;
      state.styles.popper.maxHeight = heightStrategy.max !== void 0 ? `${heightStrategy.max}px` : "100%";
    }
  },
  phase: "beforeWrite",
  requires: ["computeStyles"]
};
