// @dynatrace/strato-components/layouts/page/DetailView.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DetailView_exports = {};
__export(DetailView_exports, {
  DetailView: () => DetailView
});
module.exports = __toCommonJS(DetailView_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_intl = require("react-intl");
var styles = __toESM(require("./DetailView.sty.js"));
var import_keepMounted = require("./keepMounted.js");
var import_page_context = require("./page-context.js");
var panelStyles = __toESM(require("./panel.sty.js"));
var import_Panel = require("./Panel.js");
var import_useWatchBreakpoint = require("./useWatchBreakpoint.js");
var import_useId = require("../../core/hooks/useId.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_Drawer = require("../../overlays/drawer/Drawer.js");
const DetailView = (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const {
      breakpoint = 1280,
      className,
      defaultDismissed,
      dismissed,
      minWidth = 320,
      onDismissChange,
      onResize,
      resizable = true,
      preferredWidth = "25%",
      style,
      children,
      keepMounted = false,
      ...restProps
    } = props;
    const panelId = (0, import_useId.useId)();
    const panelName = "details";
    const panelRef = (0, import_react.useRef)(null);
    const mergedRefs = (0, import_useMergeRefs.useMergeRefs)([panelRef, forwardedRef]);
    const intl = (0, import_react_intl.useIntl)();
    const page = (0, import_page_context.usePageContext)();
    const panel = (0, import_page_context.useRegisterPagePanel)(panelName, {
      ref: panelRef,
      isControlled: dismissed !== void 0,
      isDismissed: dismissed ?? props.defaultDismissed,
      minWidth,
      breakpoint
    });
    const setIsDismissedWithReason = (0, import_react.useCallback)(
      (value, reason) => {
        if (value === true) {
          page.dismissPanel(panelName);
        } else {
          page.expandPanel(panelName);
        }
        onDismissChange?.(value, reason);
      },
      [onDismissChange, page, panelName]
    );
    const aboveBreakpoint = (0, import_useWatchBreakpoint.useWatchBreakpoint)({
      breakpoint,
      callback: setIsDismissedWithReason
    });
    const panelChildren = (0, import_keepMounted.useKeepMounted)({ children, keepMounted });
    if (!panel) {
      return null;
    }
    const label = intl.formatMessage({
      defaultMessage: "Detail View",
      id: "rR+xf1BGWvDgZVww"
    });
    const isDisplayedAsDismissed = !aboveBreakpoint || panel.isDismissed;
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      import_Panel.Panel,
      {
        name: panelName,
        id: panelId,
        className: (0, import_clsx.default)(
          isDisplayedAsDismissed && panelStyles.dismissedPanel,
          styles.details
        ),
        "aria-label": label,
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_Panel.Panel.Content,
            {
              ...restProps,
              containerSize: page.containerSize,
              ref: mergedRefs,
              label,
              onResize: (width) => onResize?.(page.getPanelSizes(), width),
              preferredWidth,
              minWidth,
              handle: resizable === true ? "left" : "hidden",
              isDismissed: isDisplayedAsDismissed,
              className: (0, import_clsx.default)(styles.content, className),
              style,
              children: panelChildren
            }
          ),
          !aboveBreakpoint ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_Drawer.Drawer,
            {
              className: panelStyles.drawer,
              isDismissed: panel.isDismissed,
              onDismiss: setIsDismissedWithReason,
              placement: "right",
              width: "wide",
              children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { ...restProps, className: panelStyles.drawerContent, children })
            }
          ) : null
        ]
      }
    );
  }
);
DetailView.displayName = "Page.DetailView";
