// @dynatrace/strato-components/layouts/page/Page.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Page_exports = {};
__export(Page_exports, {
  Page: () => Page
});
module.exports = __toCommonJS(Page_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_DetailView = require("./DetailView.js");
var import_Header = require("./Header.js");
var import_Main = require("./Main.js");
var import_page_context = require("./page-context.js");
var styles = __toESM(require("./Page.sty.js"));
var import_panel_css = require("./panel.sty.js");
var import_PanelControlButton = require("./PanelControlButton.js");
var import_Sidebar = require("./Sidebar.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useContainerSize = require("../../core/hooks/useContainerSize.js");
var import_useSlots = require("../../core/hooks/useSlots.js");
const COMPONENT_NAME = "Page";
function handleRegisterAction(state, payload) {
  const { panel, ref, isControlled, isDismissed, minWidth, breakpoint } = payload;
  if (state[panel]) {
    return state;
  }
  const newState = { ...state };
  if (panel === "main") {
    newState[panel] = {
      ref,
      isControlled: false,
      isDismissed: false,
      minWidth: minWidth ?? 0,
      breakpoint: 0
    };
    return newState;
  }
  newState[panel] = {
    ref,
    isControlled,
    isDismissed,
    minWidth: minWidth ?? 0,
    breakpoint: breakpoint ?? 0
  };
  return newState;
}
function handleRemoveAction(state, payload) {
  const { panel } = payload;
  const newState = { ...state, [panel]: null };
  return newState;
}
function handleDismissAction(state, payload) {
  const { panel } = payload;
  if (panel === "main") {
    return state;
  }
  const newState = { ...state };
  if (!newState[panel]) {
    return state;
  }
  const isControlled = newState[panel].isControlled;
  if (isControlled) {
    return state;
  }
  newState[panel] = {
    ...newState[panel],
    isDismissed: true
  };
  return newState;
}
function handleExpandAction(state, payload) {
  const { panel } = payload;
  if (panel === "main") {
    return state;
  }
  const newState = { ...state };
  if (!newState[panel]) {
    return state;
  }
  const isControlled = newState[panel].isControlled;
  if (isControlled) {
    return state;
  }
  newState[panel] = {
    ...newState[panel],
    isDismissed: false
  };
  return newState;
}
function handleToggleAction(state, payload) {
  const { panel } = payload;
  if (panel === "main") {
    return state;
  }
  const newState = { ...state };
  if (!newState[panel]) {
    return state;
  }
  const isControlled = newState[panel].isControlled;
  if (isControlled) {
    return state;
  }
  newState[panel] = {
    ...newState[panel],
    isDismissed: !newState[panel].isDismissed
  };
  return newState;
}
function pageLayoutReducer(state, action) {
  if (action.type === "register") {
    return handleRegisterAction(state, action.payload);
  }
  if (action.type === "remove") {
    return handleRemoveAction(state, action.payload);
  }
  if (action.type === "dismiss") {
    return handleDismissAction(state, action.payload);
  }
  if (action.type === "expand") {
    return handleExpandAction(state, action.payload);
  }
  if (action.type === "toggle") {
    return handleToggleAction(state, action.payload);
  }
  throw Error("invalid action.type");
}
function getGridColumnStyle(panel, max = "auto") {
  const hasDrawer = !window.matchMedia(`(min-width: ${panel.breakpoint}px)`).matches;
  if (hasDrawer || panel.minWidth === void 0 || panel.isDismissed) {
    return `minmax(min-content, ${max})`;
  }
  return `minmax(${panel.minWidth + import_panel_css.GAP_SIZE}px, ${max})`;
}
function getGridTemplateStyles(panels) {
  let columns = "";
  let contentAreas = "";
  let headerAreas = "";
  if (panels.main) {
    columns = `${getGridColumnStyle(panels.main, "1fr")} ${columns}`;
    contentAreas = "main";
    headerAreas = "header";
  }
  if (panels.sidebar) {
    columns = `${getGridColumnStyle(panels.sidebar)} ${columns}`;
    contentAreas = `sidebar ${contentAreas}`;
    headerAreas = `header ${headerAreas}`;
  }
  if (panels.details) {
    columns = `${columns} ${getGridColumnStyle(panels.details)}`;
    contentAreas = `${contentAreas} details`;
    headerAreas = `${headerAreas} header`;
  }
  return {
    gridTemplateColumns: columns,
    gridTemplateAreas: `"${headerAreas}" "${contentAreas}"`,
    gridTemplateRows: "min-content minmax(0, 1fr)"
  };
}
const PageRoot = (props) => {
  const { children, style, className, ...remainingProps } = props;
  const [slots, otherChildren] = (0, import_useSlots.useSlots)(children, {
    header: import_Header.Header
  });
  const [panels, dispatch] = (0, import_react.useReducer)(pageLayoutReducer, {
    sidebar: null,
    main: null,
    details: null
  });
  const containerRef = (0, import_react.useRef)(null);
  const containerSize = (0, import_useContainerSize.useContainerSize)(containerRef);
  const getPanelSizes = (0, import_react.useCallback)(() => {
    const sizes = {};
    for (const [key, panel] of Object.entries(panels)) {
      if (!panel?.ref.current) {
        continue;
      }
      const panelSize = parseFloat(getComputedStyle(panel.ref.current).width);
      sizes[key] = {
        px: panelSize,
        "%": panelSize / containerSize.width * 100
      };
    }
    return sizes;
  }, [containerSize.width, panels]);
  const registerPanel = (0, import_react.useCallback)(
    (panelName, { ref, isControlled, isDismissed, minWidth, breakpoint }) => {
      dispatch({
        type: "register",
        payload: {
          panel: panelName,
          ref,
          isControlled,
          isDismissed,
          minWidth,
          breakpoint
        }
      });
    },
    []
  );
  const removePanel = (0, import_react.useCallback)((panelName) => {
    dispatch({ type: "remove", payload: { panel: panelName } });
  }, []);
  const dismissPanel = (0, import_react.useCallback)((panelName) => {
    dispatch({
      type: "dismiss",
      payload: { panel: panelName }
    });
  }, []);
  const expandPanel = (0, import_react.useCallback)((panelName) => {
    dispatch({
      type: "expand",
      payload: { panel: panelName }
    });
  }, []);
  const togglePanel = (0, import_react.useCallback)((panelName) => {
    dispatch({
      type: "toggle",
      payload: { panel: panelName }
    });
  }, []);
  const gridTemplateStyles = getGridTemplateStyles(panels);
  const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const ctx = (0, import_react.useMemo)(
    () => ({
      containerSize: containerSize.width,
      panels,
      registerPanel,
      removePanel,
      dismissPanel,
      expandPanel,
      togglePanel,
      getPanelSizes
    }),
    [
      containerSize.width,
      panels,
      registerPanel,
      removePanel,
      dismissPanel,
      expandPanel,
      togglePanel,
      getPanelSizes
    ]
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    "div",
    {
      ...restProps,
      ...behavioralTrackingProps,
      ref: containerRef,
      className: (0, import_clsx.default)(styles.page, className),
      style: { ...style, ...gridTemplateStyles },
      "data-dt-component": COMPONENT_NAME,
      children: [
        slots.header,
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_page_context.pageContext.Provider, { value: ctx, children: otherChildren })
      ]
    }
  );
};
const Page = Object.assign(PageRoot, {
  /** The `Header` component is used inside the `Page` component for displaying content at the very top of the page. */
  Header: import_Header.Header,
  /** The `Sidebar` component is used inside the `Page` component for displaying content in the corresponding panel. */
  Sidebar: import_Sidebar.Sidebar,
  /**
   * The `Main` component is used inside the `Page` component for displaying content in the corresponding panel.
   * It is the only compound that is mandatory for rendering the `Page`.
   */
  Main: import_Main.Main,
  /** The `DetailView` component is used inside the `Page` component for displaying content in the corresponding panel. */
  DetailView: import_DetailView.DetailView,
  /**
   * Use the `PanelControlButton` to get the correct icon for toggling the sidebar / detail view.
   * Depending on where it is rendered and the current state of the panel, the corresponding icon is rendered.
   * You can customize the rendered content by passing any children.
   */
  PanelControlButton: import_PanelControlButton.PanelControlButton
});
