// @dynatrace/strato-components/layouts/page/page-context.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var page_context_exports = {};
__export(page_context_exports, {
  pageContext: () => pageContext,
  usePageContext: () => usePageContext,
  useRegisterPagePanel: () => useRegisterPagePanel
});
module.exports = __toCommonJS(page_context_exports);
var import_react = require("react");
const pageContext = (0, import_react.createContext)({
  containerSize: 0,
  panels: { sidebar: null, main: null, details: null },
  registerPanel: () => void 0,
  removePanel: () => void 0,
  dismissPanel: () => void 0,
  expandPanel: () => void 0,
  togglePanel: () => void 0,
  getPanelSizes: () => void 0
});
function usePageContext() {
  return (0, import_react.useContext)(pageContext);
}
function useRegisterPagePanel(name, options) {
  const { registerPanel, removePanel, panels } = usePageContext();
  const panel = panels[name];
  (0, import_react.useLayoutEffect)(() => {
    registerPanel(name, {
      ref: options.ref,
      isControlled: options.isControlled ?? false,
      isDismissed: options.isDismissed ?? false,
      minWidth: options.minWidth ?? 0,
      breakpoint: options.breakpoint ?? 0
    });
    return () => {
      removePanel(name);
    };
  }, [
    name,
    options.breakpoint,
    options.isControlled,
    options.isDismissed,
    options.minWidth,
    options.ref,
    registerPanel,
    removePanel
  ]);
  return panel;
}
