// @dynatrace/strato-components/layouts/page/PanelControlButton.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var PanelControlButton_exports = {};
__export(PanelControlButton_exports, {
  PanelControlButton: () => PanelControlButton
});
module.exports = __toCommonJS(PanelControlButton_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_page_context = require("./page-context.js");
var import_panel_context = require("./panel-context.js");
var import_Button = require("../../buttons/button/Button.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_is_node_environment = require("../../core/utils/is-node-environment.js");
const PanelControlButton = (props) => {
  const {
    action: actionProp,
    target: targetProp,
    children: childrenProp,
    onClick,
    className: consumerClassName,
    style: consumerStyle,
    ...remainingProps
  } = props;
  const intl = (0, import_react_intl.useIntl)();
  const pageCtx = (0, import_page_context.usePageContext)();
  const panelCtx = (0, import_panel_context.usePanelContext)();
  const targetPanel = targetProp ?? panelCtx?.name ?? "details";
  const action = actionProp ?? (panelCtx !== null ? "close" : "toggle");
  const panel = pageCtx.panels[targetPanel];
  const isPanelControlled = panel?.isControlled;
  const isPanelDismissed = panel?.isDismissed;
  const shouldOpen = action === "open" || action === "toggle" && isPanelDismissed === true;
  const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
  const defaultDataDtProperties = `state:${shouldOpen ? "expand" : "collapse"}`;
  const {
    "data-dt-name": dataDtName = "ExpandCollapse",
    "data-dt-properties": dataDtProperties = defaultDataDtProperties,
    ...remainingBehavioralTrackingProps
  } = behavioralTrackingProps;
  const handleClick = (event) => {
    if (isPanelControlled && !onClick && (0, import_is_node_environment.isDev)()) {
      console.warn(
        "You are using the panel control button with a controlled panel, but didn't provide an 'onClick' callback. Implement the 'onClick' to update the panel state with the panel control button."
      );
      return;
    }
    onClick?.(event);
    if (isPanelControlled) {
      return;
    }
    switch (action) {
      case "open":
        pageCtx.expandPanel(targetPanel);
        break;
      case "close":
        pageCtx.dismissPanel(targetPanel);
        break;
      case "toggle":
        pageCtx.togglePanel(targetPanel);
        break;
    }
  };
  let localizedPanel = intl.formatMessage({
    defaultMessage: "Sidebar",
    id: "MYBFw8pxHVA1QTDj",
    description: "Panel name translation"
  });
  if (targetPanel === "details") {
    localizedPanel = intl.formatMessage({
      defaultMessage: "Detail View",
      id: "0ntbampPIxNnacS8",
      description: "Panel name translation"
    });
  }
  let ariaLabel = intl.formatMessage(
    {
      defaultMessage: "Close {panelName}",
      id: "QsfELKqpB23/KKWM",
      description: "Label for the panel control button."
    },
    { panelName: localizedPanel }
  );
  if (shouldOpen) {
    ariaLabel = intl.formatMessage(
      {
        defaultMessage: "Open {panelName}",
        id: "01469lp26TkXjPEz",
        description: "Label for the panel control button."
      },
      { panelName: localizedPanel }
    );
  }
  let defaultIcon;
  if (shouldOpen) {
    defaultIcon = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.OpenSidebarIcon, {});
  } else if (action === "toggle") {
    defaultIcon = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.CloseSidebarIcon, {});
  } else {
    defaultIcon = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.XmarkIcon, {});
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Button.Button,
    {
      "aria-label": ariaLabel,
      className: consumerClassName,
      style: consumerStyle,
      onClick: handleClick,
      "data-dt-name": dataDtName,
      "data-dt-properties": dataDtProperties,
      ...remainingBehavioralTrackingProps,
      ...restProps,
      children: childrenProp ?? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        "div",
        {
          style: {
            rotate: targetPanel === "sidebar" ? "0deg" : "180deg",
            lineHeight: 0
          },
          children: defaultIcon
        }
      ) })
    }
  );
};
PanelControlButton.displayName = "Page.PanelControlButton";
