// @dynatrace/strato-components/layouts/page/Panel.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Panel_exports = {};
__export(Panel_exports, {
  Panel: () => Panel
});
module.exports = __toCommonJS(Panel_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("@use-gesture/react");
var import_clsx = __toESM(require("clsx"));
var import_react2 = require("react");
var import_react_intl = require("react-intl");
var import_keepMounted = require("./keepMounted.js");
var import_panel_context = require("./panel-context.js");
var styles = __toESM(require("./panel.sty.js"));
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_useFocusRing = require("../../core/styles/useFocusRing.js");
const PanelRoot = (props) => {
  const { className, id, style, children, name } = props;
  const ctx = (0, import_react2.useMemo)(() => ({ name }), [name]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_panel_context.panelContext.Provider, { value: ctx, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_keepMounted.KeepMountedProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "aside",
    {
      ...props,
      id,
      className: (0, import_clsx.default)(styles.root, className),
      style,
      children
    }
  ) }) });
};
const PanelContent = (0, import_react2.forwardRef)(
  (props, forwardedRef) => {
    const {
      children,
      className,
      handle = "right",
      isDismissed,
      label,
      minWidth,
      onResize,
      preferredWidth,
      style,
      containerSize,
      ...restProps
    } = props;
    const panelRef = (0, import_react2.useRef)(null);
    const panelRefs = (0, import_useMergeRefs.useMergeRefs)([panelRef, forwardedRef]);
    const handleRef = (0, import_react2.useRef)(null);
    const draggedSize = (0, import_react2.useRef)();
    const wasResized = (0, import_react2.useRef)(false);
    const setPanelSize = (0, import_react2.useCallback)(
      (size) => {
        if (!panelRef.current) {
          return;
        }
        let sizeToApply = Math.floor(size * 1e3) / 1e3;
        draggedSize.current = sizeToApply;
        if (size < 0) {
          return;
        }
        if (minWidth && sizeToApply < minWidth) {
          sizeToApply = minWidth;
        }
        panelRef.current.style.setProperty("width", `${sizeToApply}px`);
      },
      [minWidth]
    );
    const ensurePanelSizeMatchesComputedWidth = (0, import_react2.useCallback)(
      (element, size) => {
        const actualWidth = parseFloat(getComputedStyle(element).width);
        if (size !== actualWidth) {
          setPanelSize(actualWidth);
        }
        return actualWidth;
      },
      [setPanelSize]
    );
    const handleDrag = (0, import_react2.useCallback)(
      (deltaX) => {
        if (!draggedSize.current) {
          return;
        }
        setPanelSize(
          draggedSize.current + deltaX * (handle === "right" ? 1 : -1)
        );
        wasResized.current = true;
      },
      [handle, setPanelSize]
    );
    const handleDragEnd = (0, import_react2.useCallback)(() => {
      if (!panelRef.current || !draggedSize.current) {
        return;
      }
      const size = ensurePanelSizeMatchesComputedWidth(
        panelRef.current,
        draggedSize.current
      );
      onResize?.(size);
    }, [ensurePanelSizeMatchesComputedWidth, onResize]);
    const handleDragStart = (0, import_react2.useCallback)(() => {
      if (!panelRef.current || !draggedSize.current) {
        return;
      }
      ensurePanelSizeMatchesComputedWidth(
        panelRef.current,
        draggedSize.current
      );
    }, [ensurePanelSizeMatchesComputedWidth]);
    (0, import_react.useGesture)(
      {
        onDragStart: handleDragStart,
        onDragEnd: handleDragEnd,
        onDrag: ({ delta: [deltaX] }) => handleDrag(deltaX)
      },
      {
        target: handleRef,
        eventOptions: { passive: false },
        drag: { filterTaps: true }
      }
    );
    (0, import_react2.useEffect)(
      function initializePanelSize() {
        if (!panelRef.current || wasResized.current || isDismissed) {
          return;
        }
        if (typeof preferredWidth === "number") {
          setPanelSize(preferredWidth);
          return;
        }
        if (preferredWidth?.includes("%")) {
          const size2 = containerSize * (parseFloat(preferredWidth) / 100);
          setPanelSize(size2);
          return;
        }
        if (preferredWidth?.includes("px")) {
          const size2 = parseFloat(preferredWidth);
          setPanelSize(size2);
          return;
        }
        if (typeof preferredWidth === "string") {
          panelRef.current.style.setProperty("width", preferredWidth);
          const size2 = parseFloat(getComputedStyle(panelRef.current).width);
          setPanelSize(size2);
          return;
        }
        const size = parseFloat(getComputedStyle(panelRef.current).width);
        setPanelSize(size);
      },
      [containerSize, preferredWidth, setPanelSize, isDismissed]
    );
    if (isDismissed) {
      return null;
    }
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        "div",
        {
          ...restProps,
          ref: panelRefs,
          className: (0, import_clsx.default)(styles.panelContent, className),
          style: { ...style, minWidth, width: draggedSize.current },
          children
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: styles.resizeHandleContainer, children: handle !== "hidden" ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        ResizeHandle,
        {
          panelName: label,
          panelSize: draggedSize.current,
          ref: handleRef
        }
      ) : null })
    ] });
  }
);
const ResizeHandle = (0, import_react2.forwardRef)(
  (props, ref) => {
    const { panelName, panelSize = 0 } = props;
    const intl = (0, import_react_intl.useIntl)();
    const { focusProps, focusClassName } = (0, import_useFocusRing.useFocusRing)({
      focusWithin: true,
      ignoreModality: true
    });
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "hr",
      {
        ...focusProps,
        ref,
        className: (0, import_clsx.default)(styles.resizeHandle, focusClassName),
        "aria-valuenow": panelSize,
        "aria-valuetext": `${Math.round(panelSize)}px`,
        "aria-orientation": "vertical",
        "aria-label": intl.formatMessage(
          {
            defaultMessage: "Resize {label} panel",
            id: "Ym3RUHtaLVFsGlaM",
            description: "Label for the resize handle of the panel to resize it with drag and drop or keyboard."
          },
          { label: panelName }
        ),
        tabIndex: 0,
        "data-dt-name": "Resize"
      }
    );
  }
);
const Panel = Object.assign(PanelRoot, {
  Content: PanelContent
});
