// @dynatrace/strato-components/layouts/page/panel.sty.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var panel_css_exports = {};
__export(panel_css_exports, {
  GAP_SIZE: () => GAP_SIZE,
  buttonIcon: () => buttonIcon,
  dismissedPanel: () => dismissedPanel,
  drawer: () => drawer,
  drawerContent: () => drawerContent,
  gap: () => gap,
  panelContent: () => panelContent,
  resizeHandle: () => resizeHandle,
  resizeHandleContainer: () => resizeHandleContainer,
  root: () => root,
  rtlPanel: () => rtlPanel
});
module.exports = __toCommonJS(panel_css_exports);
var import_AppRoot_css_ts_vanilla = require("../../core/components/app-root/AppRoot.css");
var import_panel_css_ts_vanilla = require("./panel.css");
var GAP_SIZE = 8;
var buttonIcon = "_1p91g306-3-6-0";
var dismissedPanel = "_1p91g302-3-6-0";
var drawer = "_1p91g304-3-6-0";
var drawerContent = "_1p91g305-3-6-0";
var gap = "_1p91g303-3-6-0";
var panelContent = "_1p91g307-3-6-0";
var resizeHandle = "_1p91g309-3-6-0";
var resizeHandleContainer = "_1p91g308-3-6-0";
var root = "_1p91g301-3-6-0";
var rtlPanel = "_1p91g300-3-6-0";
