// @dynatrace/strato-components/layouts/page/Sidebar.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Sidebar_exports = {};
__export(Sidebar_exports, {
  IntegratedControls: () => IntegratedControls,
  Sidebar: () => Sidebar
});
module.exports = __toCommonJS(Sidebar_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_intl = require("react-intl");
var import_strato_icons = require("@dynatrace/strato-icons");
var import_keepMounted = require("./keepMounted.js");
var import_page_context = require("./page-context.js");
var panelStyles = __toESM(require("./panel.sty.js"));
var import_Panel = require("./Panel.js");
var styles = __toESM(require("./Sidebar.sty.js"));
var import_useWatchBreakpoint = require("./useWatchBreakpoint.js");
var import_Button = require("../../buttons/button/Button.js");
var import_useId = require("../../core/hooks/useId.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_Drawer = require("../../overlays/drawer/Drawer.js");
var import_Tooltip = require("../../overlays/tooltip/Tooltip.js");
const Sidebar = (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const {
      breakpoint = 640,
      className,
      defaultDismissed,
      dismissed,
      minWidth = 200,
      onDismissChange,
      onResize,
      preferredWidth = "15%",
      style,
      resizable = true,
      integratedControls,
      children,
      keepMounted = false,
      ...restProps
    } = props;
    const panelId = (0, import_useId.useId)();
    const panelName = "sidebar";
    const panelRef = (0, import_react.useRef)(null);
    const mergedRefs = (0, import_useMergeRefs.useMergeRefs)([panelRef, forwardedRef]);
    const intl = (0, import_react_intl.useIntl)();
    const page = (0, import_page_context.usePageContext)();
    const panel = (0, import_page_context.useRegisterPagePanel)(panelName, {
      ref: panelRef,
      isControlled: dismissed !== void 0,
      isDismissed: dismissed ?? props.defaultDismissed,
      minWidth,
      breakpoint
    });
    const setIsDismissedWithReason = (0, import_react.useCallback)(
      (value, reason) => {
        if (value === true) {
          page.dismissPanel(panelName);
        } else {
          page.expandPanel(panelName);
        }
        onDismissChange?.(value, reason);
      },
      [onDismissChange, page, panelName]
    );
    const aboveBreakpoint = (0, import_useWatchBreakpoint.useWatchBreakpoint)({
      breakpoint,
      callback: setIsDismissedWithReason
    });
    const panelChildren = (0, import_keepMounted.useKeepMounted)({ children, keepMounted });
    if (!panel) {
      return null;
    }
    const label = intl.formatMessage({
      defaultMessage: "Sidebar",
      id: "9RcckxxccNTcNVf9"
    });
    const isDisplayedAsDismissed = !aboveBreakpoint || panel.isDismissed;
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      import_Panel.Panel,
      {
        name: panelName,
        id: panelId,
        className: (0, import_clsx.default)(
          isDisplayedAsDismissed && panelStyles.dismissedPanel,
          integratedControls && styles.integratedControlsPanel,
          styles.sidebar
        ),
        "aria-label": label,
        children: [
          isDisplayedAsDismissed ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: panelStyles.gap }) : null,
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_Panel.Panel.Content,
            {
              ...restProps,
              containerSize: page.containerSize,
              ref: mergedRefs,
              label,
              onResize: (width) => onResize?.(page.getPanelSizes(), width),
              preferredWidth,
              minWidth,
              handle: resizable === true ? "right" : "hidden",
              isDismissed: isDisplayedAsDismissed,
              className,
              style,
              children: panelChildren
            }
          ),
          integratedControls ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            IntegratedControls,
            {
              dismissed: isDisplayedAsDismissed,
              onClick: ({ value, reason }) => {
                setIsDismissedWithReason(value, reason);
              },
              panelId,
              label
            }
          ) : null,
          !aboveBreakpoint ? /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
            import_Drawer.Drawer,
            {
              className: panelStyles.drawer,
              isDismissed: panel.isDismissed,
              onDismiss: setIsDismissedWithReason,
              placement: "left",
              width: "wide",
              children: [
                integratedControls ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: styles.integratedControlsDrawerHeader, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                  import_Button.Button,
                  {
                    style: { marginLeft: "auto" },
                    onClick: () => setIsDismissedWithReason(true, "drawer button"),
                    children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.XmarkIcon, {}) })
                  }
                ) }) : null,
                /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { ...restProps, className: panelStyles.drawerContent, children })
              ]
            }
          ) : null
        ]
      }
    );
  }
);
Sidebar.displayName = "Page.Sidebar";
const IntegratedControls = ({
  dismissed,
  onClick,
  panelId,
  label
}) => {
  const intl = (0, import_react_intl.useIntl)();
  const buttonRef = (0, import_react.useRef)(null);
  const [isTooltipOpen, setIsTooltipOpen] = (0, import_react.useState)(false);
  const labelText = intl.formatMessage(
    dismissed ? { defaultMessage: "Expand {label} panel", id: "TWLhm3I5du5FpaM+" } : { defaultMessage: "Dismiss {label} panel", id: "SWpMc7FyaE8Uxhzn" },
    { label }
  );
  const buttonProps = dismissed ? {
    "aria-expanded": "false",
    className: styles.expandButton,
    onClick: () => {
      onClick({ value: false, reason: "expand button" });
      setIsTooltipOpen(false);
    },
    variant: "emphasized"
  } : {
    "aria-expanded": "true",
    className: styles.dismissButton,
    onClick: () => {
      onClick({ value: true, reason: "dismiss button" });
      setIsTooltipOpen(false);
    },
    variant: "default"
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { className: styles.integratedControls, children: [
    dismissed ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "button",
      {
        className: styles.hint,
        type: "button",
        onClick: () => {
          onClick({ value: false, reason: "panel hint" });
          setIsTooltipOpen(false);
          buttonRef.current?.focus();
        },
        "aria-controls": panelId,
        "aria-expanded": dismissed ? "false" : "true",
        "aria-label": labelText
      }
    ) : null,
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: styles.buttonContainer, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_Tooltip.Tooltip,
      {
        text: labelText,
        open: isTooltipOpen,
        onOpenChange: setIsTooltipOpen,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_Button.Button,
          {
            ...buttonProps,
            ref: buttonRef,
            "aria-controls": panelId,
            "aria-label": labelText,
            size: "condensed",
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Button.Button.Prefix, { children: dismissed ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ChevronSmallRightDoubleIcon, {}) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_strato_icons.ChevronSmallLeftDoubleIcon, {}) })
          }
        )
      }
    ) })
  ] });
};
