// @dynatrace/strato-components/layouts/page/Sidebar.sty.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Sidebar_css_exports = {};
__export(Sidebar_css_exports, {
  buttonContainer: () => buttonContainer,
  dismissButton: () => dismissButton,
  expandButton: () => expandButton,
  hint: () => hint,
  integratedControls: () => integratedControls,
  integratedControlsDrawerHeader: () => integratedControlsDrawerHeader,
  integratedControlsPanel: () => integratedControlsPanel,
  sidebar: () => sidebar
});
module.exports = __toCommonJS(Sidebar_css_exports);
var import_AppRoot_css_ts_vanilla = require("../../core/components/app-root/AppRoot.css");
var import_panel_css_ts_vanilla = require("./panel.css");
var import_focusRingStatic_css_ts_vanilla = require("../../core/styles/focusRingStatic.css");
var import_Sidebar_css_ts_vanilla = require("./Sidebar.css");
var buttonContainer = "_zdcge95-3-6-0";
var dismissButton = "_zdcge97-3-6-0 _zdcge96-3-6-0";
var expandButton = "_zdcge98-3-6-0 _zdcge96-3-6-0";
var hint = "_zdcge93-3-6-0 _z0fq9b4-3-6-0 _z0fq9bd-3-6-0 _z0fq9b6-3-6-0 _z0fq9bc-3-6-0";
var integratedControls = "_zdcge91-3-6-0";
var integratedControlsDrawerHeader = "_zdcge9a-3-6-0";
var integratedControlsPanel = "_zdcge99-3-6-0";
var sidebar = "_zdcge90-3-6-0";
