// @dynatrace/strato-components/charts/pie/components/PieChartTooltip/PieChartTooltip.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var PieChartTooltip_exports = {};
__export(PieChartTooltip_exports, {
  PieChartTooltip: () => PieChartTooltip
});
module.exports = __toCommonJS(PieChartTooltip_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_OverlayTooltip = require("../../../core/components/overlay-tooltip/OverlayTooltip.js");
var import_selectors = require("../../../core/components/overlay-tooltip/selectors.js");
var import_utils = require("../../../core/components/overlay-tooltip/utils.js");
var import_useTooltipSeriesActions = require("../../../core/components/tooltip/shared/hooks/useTooltipSeriesActions.js");
var import_selectors2 = require("../../hooks/store/selectors.js");
var import_usePieChartTooltip = require("../../hooks/usePieChartTooltip.js");
const PieChartTooltip = () => {
  const customTooltip = (0, import_usePieChartTooltip.usePieChartTooltip)();
  const overlayTooltipPinned = (0, import_selectors.useSelector)(
    (state) => state.pinned
  );
  const overlayData = (0, import_selectors.useSelector)((state) => state.data);
  const visibleSlices = (0, import_selectors2.useVisibleSlicesState)();
  const seriesActions = (0, import_useTooltipSeriesActions.useTooltipSeriesActions)(
    visibleSlices,
    (slice) => slice.id
  );
  if (customTooltip?.hidden || (0, import_utils.isTooltipDataEmpty)(overlayData)) {
    return null;
  }
  const metadata = overlayData.metadata;
  const currentSeries = overlayData.series[0];
  const fallbackSeriesId = typeof currentSeries.name === "string" ? currentSeries.name : "";
  if (customTooltip?.template) {
    const payload = {
      category: typeof currentSeries.name === "string" ? currentSeries.name : "",
      value: currentSeries.value,
      percentage: currentSeries.relativeValue,
      color: currentSeries.color
    };
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayTooltip.OverlayTooltip, { legacyTemplate: customTooltip.template(payload) });
  }
  const tooltipItem = currentSeries ? {
    id: metadata?.sliceId ?? fallbackSeriesId,
    name: currentSeries.name,
    color: { Default: currentSeries.color },
    values: [
      currentSeries.value?.toString() || "",
      currentSeries.relativeValue?.toString() || ""
    ]
  } : void 0;
  const actions = overlayTooltipPinned && tooltipItem ? () => seriesActions(tooltipItem) : void 0;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OverlayTooltip.OverlayTooltip, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_OverlayTooltip.OverlayTooltip.SeriesItem,
    {
      name: currentSeries.name,
      value: currentSeries.value,
      color: currentSeries.color,
      relativeValue: currentSeries.relativeValue,
      relativeValueUnit: currentSeries.relativeValueUnit,
      actions
    }
  ) });
};
