// @dynatrace/strato-components/charts/xy-chart/processing-strategies/utils/access-datapoint/get-datapoints.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var get_datapoints_exports = {};
__export(get_datapoints_exports, {
  getAxisType: () => getAxisType,
  getBaseDatapoint: () => getBaseDatapoint,
  getChartAreaVariant: () => getChartAreaVariant,
  getChartOrientation: () => getChartOrientation,
  getDatumValue: () => getDatumValue,
  getDotDatapoint: () => getDotDatapoint,
  getLineDatapoint: () => getLineDatapoint,
  getPoint: () => getPoint,
  getRectDatapoint: () => getRectDatapoint
});
module.exports = __toCommonJS(get_datapoints_exports);
var import_lodash_es = require("lodash");
var import_axis = require("../../../types/axis.js");
var import_guards = require("../../../types/guards.js");
var import_cast_data_types = require("../cast-data/cast-data-types.js");
function getDatumValue(datum, path) {
  if (!path) {
    return void 0;
  }
  if ((0, import_lodash_es.isFunction)(path)) {
    return path(datum);
  }
  return (0, import_lodash_es.get)(datum, path);
}
function getPoint(datum, path, axisType) {
  const value = getDatumValue(datum, path);
  return (0, import_cast_data_types.castDataType)(value, axisType);
}
function getBaseDatapoint(datum, seriesProps, xAxis, yAxis) {
  const x0 = getPoint(datum, seriesProps.x0Accessor, xAxis.type);
  const y0 = getPoint(datum, seriesProps.y0Accessor, yAxis.type);
  let x1;
  let y1;
  if ((0, import_lodash_es.isUndefined)(x0) || (0, import_lodash_es.isUndefined)(y0)) {
    return;
  }
  if ((0, import_guards.isTime)(xAxis.type) && seriesProps.x1Accessor) {
    x1 = getPoint(datum, seriesProps.x1Accessor, import_axis.XYAxisTypeTime);
  }
  if ((0, import_guards.isTime)(yAxis.type) && seriesProps.y1Accessor) {
    y1 = getPoint(datum, seriesProps.y1Accessor, import_axis.XYAxisTypeTime);
  }
  return { ...datum, x0, x1, y0, y1 };
}
function getDotDatapoint(datum, seriesProps, xAxis, yAxis) {
  return getBaseDatapoint(datum, seriesProps, xAxis, yAxis);
}
function getRectDatapoint(datum, seriesProps, xAxis, yAxis) {
  const x0 = getPoint(datum, seriesProps.x0Accessor, xAxis.type);
  if (x0 === void 0) {
    return void 0;
  }
  const x1 = seriesProps.x1Accessor && getPoint(datum, seriesProps.x1Accessor, xAxis.type);
  if (x1 === void 0 && !(0, import_guards.isCategorical)(xAxis.type)) {
    return void 0;
  }
  const y0 = getPoint(datum, seriesProps.y0Accessor, yAxis.type);
  if (y0 === void 0) {
    return void 0;
  }
  const y1 = seriesProps.y1Accessor && getPoint(datum, seriesProps.y1Accessor, yAxis.type);
  if (y1 === void 0 && !(0, import_guards.isCategorical)(yAxis.type)) {
    return void 0;
  }
  const value = getDatumValue(
    datum,
    seriesProps.valueAccessor
  );
  return {
    ...datum,
    value,
    x0,
    x1,
    y0,
    y1
  };
}
function getLineDatapoint(datum, seriesProps, xAxis, yAxis) {
  return getBaseDatapoint(datum, seriesProps, xAxis, yAxis);
}
const getAxisType = (axisName, xAxis, yAxis) => {
  const axisType = {
    x: xAxis,
    y: yAxis
  };
  return axisType[axisName].type;
};
const getChartOrientation = ({
  x1Accessor,
  x0Accessor,
  y1Accessor,
  y0Accessor
}) => {
  const isVertical = !!(x0Accessor && y1Accessor);
  const isHorizontal = !!(x1Accessor && y0Accessor);
  const isFullSquare = isVertical && isHorizontal;
  if (isFullSquare) {
    return "both";
  }
  if (isVertical) {
    return "vertical";
  }
  if (isHorizontal) {
    return "horizontal";
  } else {
    return void 0;
  }
};
const getChartAreaVariant = (chartOrientation, seriesProps) => {
  if (chartOrientation === "both") {
    return "band";
  }
  if (chartOrientation === "horizontal") {
    return seriesProps.x0Accessor ? "band" : "area";
  }
  return seriesProps.y0Accessor ? "band" : "area";
};
