// @dynatrace/strato-components/core/components/time-indicator/TimeIndicator.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TimeIndicator_exports = {};
__export(TimeIndicator_exports, {
  TimeIndicator: () => TimeIndicator
});
module.exports = __toCommonJS(TimeIndicator_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_tz = require("@date-fns/tz");
var import_react_intl = require("react-intl");
var import_user_preferences = require("@dynatrace-sdk/user-preferences");
var import_useIntervalCallback = require("../../time/useIntervalCallback.js");
const MilliSecondsInfo = {
  OneMinute: 6e4,
  OneHour: 36e5,
  OneDay: 864e5
};
const TimeIndicator = (props) => {
  const {
    time,
    "data-testid": dataTestId,
    className: consumerClassName,
    style: consumerStyle,
    "data-dtrum-mask": dataDtrumMask,
    "data-dtrum-allow": dataDtrumAllow,
    ...remainingProps
  } = props;
  const date = import_tz.TZDateMini.tz((0, import_user_preferences.getTimezone)());
  const { value: now } = (0, import_useIntervalCallback.useIntervalCallback)(
    () => date.getTime(),
    [],
    (time - date.getSeconds() * 1e3) % MilliSecondsInfo.OneMinute,
    MilliSecondsInfo.OneMinute
  );
  const currentTime = now ?? date.getTime();
  const diff = currentTime - time;
  let content;
  if (diff < MilliSecondsInfo.OneMinute) {
    content = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_react_intl.FormattedMessage,
      {
        defaultMessage: "Just now",
        id: "HPbv8ZQrbBL7hRCT",
        description: "Just now marker in the FilterField recent suggestions"
      }
    );
  } else if (diff < MilliSecondsInfo.OneHour) {
    const minutes = -Math.floor(diff / MilliSecondsInfo.OneMinute);
    content = // eslint rule is not smart in checking this. The style prop comes from the lib.Intl types, where it defines the style of the formatted value.
    // eslint-disable-next-line react/style-prop-object
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_intl.FormattedRelativeTime, { value: minutes, unit: "minute", style: "narrow" });
  } else if (diff < MilliSecondsInfo.OneDay) {
    const hours = -Math.floor(diff / MilliSecondsInfo.OneHour);
    content = // eslint rule is not smart in checking this. The style prop comes from the lib.Intl types, where it defines the style of the formatted value.
    // eslint-disable-next-line react/style-prop-object
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_intl.FormattedRelativeTime, { value: hours, unit: "hour", style: "narrow" });
  } else {
    const days = -Math.floor(diff / MilliSecondsInfo.OneDay);
    content = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_react_intl.FormattedRelativeTime, { value: days, unit: "day", style: "narrow" });
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "span",
    {
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...remainingProps,
      children: content
    }
  );
};
