// @dynatrace/strato-components/esm/filters/timeframe-selector/components/stepper/StepperButtonTooltip.js
import { jsx, jsxs } from "react/jsx-runtime";
import { Flex } from "../../../../layouts/flex/Flex.js";
import { Tooltip } from "../../../../overlays/tooltip/Tooltip.js";
import { Text } from "../../../../typography/text/Text.js";
import { useTooltipMessages } from "../../hooks/useTooltipMessages.js";
const StepperButtonTooltip = (props) => {
  const {
    value,
    disabled,
    children,
    formattedStepperValue
  } = props;
  const { tooltipFormattedTimeValueStart, tooltipFormattedTimeValueEnd } = useTooltipMessages({
    from: value?.from.value ?? "",
    to: value?.to.value ?? ""
  });
  return /* @__PURE__ */ jsx(
    Tooltip,
    {
      text: /* @__PURE__ */ jsxs(Flex, { flexDirection: "column", gap: 0, children: [
        /* @__PURE__ */ jsx(Text, { textStyle: "small-emphasized", children: formattedStepperValue }),
        /* @__PURE__ */ jsx(Text, { textStyle: "small", children: tooltipFormattedTimeValueStart }),
        /* @__PURE__ */ jsx(Text, { textStyle: "small", children: tooltipFormattedTimeValueEnd })
      ] }),
      placement: "top",
      fallbackPlacements: ["top-end", "bottom", "bottom-end", "right", "left"],
      disabled,
      children
    }
  );
};
StepperButtonTooltip.displayName = "TimeframeStepperTooltip";
export {
  StepperButtonTooltip
};
