// @dynatrace/strato-components/esm/filters/timeframe-selector/components/stepper/StepperButton.js
import { jsx } from "react/jsx-runtime";
import { useCallback, useEffect, useState } from "react";
import { useIntl } from "react-intl";
import {
  ChevronLeftSmallIcon,
  ChevronRightSmallIcon
} from "@dynatrace/strato-icons";
import { StepperButtonTooltip } from "./StepperButtonTooltip.js";
import { Button } from "../../../../buttons/button/Button.js";
import { isFullTimeframe } from "../../../../core/time/helpers.js";
import { useTimeframeSelectorStore } from "../../../../core/time/timeframeStore/useTimeframeStore.js";
import { useTimeframeSelectorContext } from "../../hooks/useTimeframeSelectorContext.js";
import {
  formatStepperValues,
  getTimeShift
} from "../../utils/stepper-helpers.js";
import {
  dateIsMinMax,
  isBiggerThan,
  isSmallerThan
} from "../../utils/timeframe-selection.js";
function isStepperDisabled(direction, value, min, max) {
  if (!value) {
    return false;
  }
  if (direction === "backward") {
    const nextStepIsValid = isSmallerThan(value.from.absoluteDate, min) || dateIsMinMax(value.from, min);
    return nextStepIsValid;
  }
  if (direction === "forward") {
    const nextStepIsValid = isBiggerThan(value.to.absoluteDate, max) || dateIsMinMax(value.to, void 0, max);
    return nextStepIsValid;
  }
  return false;
}
const StepperButton = (props) => {
  const {
    disabled,
    direction,
    "aria-disabled": ariaDisabledProp,
    onBlur,
    readOnly
  } = props;
  const { showError } = useTimeframeSelectorContext();
  const precision = useTimeframeSelectorStore((state) => state.precision);
  const min = useTimeframeSelectorStore((state) => state.min)?.value;
  const max = useTimeframeSelectorStore((state) => state.max)?.value;
  const isFullDay = useTimeframeSelectorStore((state) => state.isFullDay);
  const setTimeframeSelectorTimeframe = useTimeframeSelectorStore(
    (state) => state.setStepperValue
  );
  const value = useTimeframeSelectorStore(
    (state) => isFullTimeframe(state.timeframeSelectorTimeframe) ? state.timeframeSelectorTimeframe : null
  );
  const referenceTimeframe = useTimeframeSelectorStore(
    (state) => state.referenceTimeframe
  );
  const setReferenceTimeframe = useTimeframeSelectorStore(
    (state) => state.setReferenceTimeframe
  );
  const [shiftValue, setShiftValue] = useState(value);
  const [isDisabled, setIsDisabled] = useState(
    isStepperDisabled(direction, value, min, max)
  );
  useEffect(() => {
    setIsDisabled(isStepperDisabled(direction, value, min, max));
    const newShiftValue = getTimeShift(value, {
      direction,
      min,
      max,
      selectedOverCalendar: isFullDay,
      precision
    });
    if (newShiftValue) {
      setShiftValue(newShiftValue);
    }
  }, [direction, value, min, max, isFullDay, isDisabled, precision]);
  const handleShift = useCallback(() => {
    const newShiftValue = getTimeShift(referenceTimeframe ?? value, {
      direction,
      min,
      max,
      selectedOverCalendar: isFullDay,
      precision
    });
    if (newShiftValue && !readOnly) {
      const isAtMinOrMax = direction === "backward" ? dateIsMinMax(newShiftValue?.from, min, max) : dateIsMinMax(newShiftValue?.to, min, max);
      if (referenceTimeframe && isFullTimeframe(referenceTimeframe)) {
        setTimeframeSelectorTimeframe(referenceTimeframe);
        setReferenceTimeframe(null);
        return;
      }
      if (isAtMinOrMax) {
        setReferenceTimeframe(value);
      }
      setTimeframeSelectorTimeframe(newShiftValue);
    }
  }, [
    value,
    direction,
    min,
    max,
    isFullDay,
    precision,
    readOnly,
    referenceTimeframe,
    setReferenceTimeframe,
    setTimeframeSelectorTimeframe
  ]);
  const intl = useIntl();
  const dataDtName = direction === "backward" ? "StepBackward" : "StepForward";
  if (!value || !shiftValue) {
    return null;
  }
  const formattedStepperValue = formatStepperValues(
    shiftValue.from,
    shiftValue.to,
    intl,
    direction,
    isFullDay
  );
  return /* @__PURE__ */ jsx(
    StepperButtonTooltip,
    {
      disabled: disabled || isDisabled,
      value: shiftValue,
      formattedStepperValue,
      children: /* @__PURE__ */ jsx(
        Button,
        {
          style: { minWidth: "fit-content" },
          variant: "emphasized",
          color: showError ? "critical" : "neutral",
          disabled: disabled || isDisabled,
          onClick: handleShift,
          "aria-disabled": ariaDisabledProp,
          "aria-label": formattedStepperValue,
          onBlur,
          "data-dt-name": dataDtName,
          children: direction === "backward" ? /* @__PURE__ */ jsx(ChevronLeftSmallIcon, {}) : /* @__PURE__ */ jsx(ChevronRightSmallIcon, {})
        }
      )
    }
  );
};
StepperButton.displayName = "TimeframeStepperButton";
export {
  StepperButton
};
