// @dynatrace/strato-components/charts/core/components/overlay-tooltip/hooks/useDynamicTooltip.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useDynamicTooltip_exports = {};
__export(useDynamicTooltip_exports, {
  useDynamicTooltip: () => useDynamicTooltip
});
module.exports = __toCommonJS(useDynamicTooltip_exports);
var import_react = require("react");
var import_react_popper = require("react-popper");
var import_useOverlayStacking = require("../../../../../core/hooks/useOverlayStacking.js");
var import_create_dom_rect = require("../../../../../core/utils/create-dom-rect.js");
const OPTIONS = {
  strategy: "fixed",
  placement: "right-start",
  modifiers: [
    {
      name: "arrow",
      enabled: false
    },
    {
      name: "offset",
      options: {
        offset: [-10, 10]
      }
    },
    {
      name: "eventListeners",
      enabled: false
    },
    {
      name: "flip",
      options: {
        padding: 10,
        flipVariations: false
      }
    }
  ]
};
const useDynamicTooltip = (args) => {
  const { template, x, y } = args;
  const virtualElement = (0, import_react.useRef)({
    x: -1,
    y: -1,
    getBoundingClientRect: (0, import_create_dom_rect.createDOMRect)({ x: -1, y: -1 })
  });
  const overlayStacking = (0, import_useOverlayStacking.useOverlayStacking)(0, "overlay");
  if (x !== virtualElement.current.x || y !== virtualElement.current.y) {
    virtualElement.current = {
      x,
      y,
      getBoundingClientRect: (0, import_create_dom_rect.createDOMRect)({ x, y })
    };
  }
  const { styles, attributes } = (0, import_react_popper.usePopper)(
    virtualElement.current,
    template,
    OPTIONS
  );
  return {
    style: {
      ...styles.popper,
      zIndex: `${overlayStacking.style.zIndex}`
    },
    ...attributes.popper
  };
};
