// @dynatrace/strato-components/charts/pie/components/PieChartPlot/PieChartPlot.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var PieChartPlot_exports = {};
__export(PieChartPlot_exports, {
  PieChartPlot: () => PieChartPlot
});
module.exports = __toCommonJS(PieChartPlot_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_group = require("@visx/group");
var import_shape = require("@visx/shape");
var import_d3_shape = require("d3-shape");
var import_react = require("react");
var import_PieChartPlot_css = require("./PieChartPlot.sty.js");
var import_useGraphSize = require("../../../core/components/chart-layout/hooks/useGraphSize.js");
var import_useTooltipReducer = require("../../../core/components/overlay-tooltip/useTooltipReducer.js");
var import_constants = require("../../constants.js");
var import_FormatterContext = require("../../context/FormatterContext.js");
var import_LabelsContext = require("../../context/LabelsContext.js");
var import_selectors = require("../../hooks/store/selectors.js");
var import_build_chart_sizes = require("../../utils/build-chart-sizes.js");
var import_build_label = require("../../utils/build-label.js");
var import_label_collisions = require("../../utils/label-collisions.js");
var import_InnerWrapper = require("../InnerWrapper/InnerWrapper.js");
var import_PieChartLabel = require("../PieChartLabel.js");
var import_PieChartSlice = require("../PieChartSlice.js");
const getValue = (pieData) => pieData.relativeValue;
const PieChartPlot = (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const { variant } = props;
    const dispatchReducer = (0, import_useTooltipReducer.useTooltipReducer)();
    const visibleSlices = (0, import_selectors.useVisibleSlicesState)();
    const highlightedId = (0, import_selectors.useHighlightedIdState)();
    const { height, width } = (0, import_useGraphSize.useGraphSize)();
    const { hidden: hiddenLabels, valueType } = (0, import_LabelsContext.useLabels)();
    const formatter = (0, import_FormatterContext.useFormatter)();
    const truncationMode = (0, import_FormatterContext.useTruncationMode)();
    const { topCenter, radius, leftCenter, innerRadius } = (0, import_build_chart_sizes.buildChartSizes)(
      height,
      width,
      variant,
      visibleSlices,
      !!hiddenLabels
    );
    const visibleIds = visibleSlices.map((slice) => slice.id);
    const donutOverlay = (0, import_react.useMemo)(
      () => (0, import_d3_shape.arc)().innerRadius(innerRadius).outerRadius(radius).startAngle(0).endAngle(360)({}),
      [innerRadius, radius]
    );
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { className: import_PieChartPlot_css.container, ref: forwardedRef, children: [
      variant === import_constants.CHART_VARIANT.Donut && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_InnerWrapper.InnerWrapper,
        {
          width,
          height,
          innerRadius,
          visibleIds
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("svg", { width, height, style: { overflow: "visible" }, children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
        import_group.Group,
        {
          top: topCenter,
          left: leftCenter,
          onMouseLeave: dispatchReducer({ type: "SET_OUT_BOUNDS" }),
          children: [
            variant === import_constants.CHART_VARIANT.Pie ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              "circle",
              {
                cx: "0",
                cy: "0",
                r: radius > 0 ? radius : 0,
                fill: "transparent"
              }
            ) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)("path", { d: donutOverlay, fill: "transparent" }),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_shape.Pie,
              {
                data: visibleSlices,
                pieValue: getValue,
                outerRadius: radius,
                innerRadius,
                padAngle: import_constants.PAD_ANGLE,
                pieSort: null,
                children: (pie) => {
                  let previousLabel = void 0;
                  return pie.arcs.map((arc2) => {
                    const label = (0, import_build_label.buildLabel)(
                      arc2,
                      pie.path,
                      valueType,
                      formatter,
                      truncationMode,
                      leftCenter
                    );
                    const isHighlighted = !highlightedId || highlightedId === arc2.data.id;
                    const showLabel = !hiddenLabels && (0, import_label_collisions.isInBounds)(label, leftCenter, topCenter) && (!previousLabel || !(0, import_label_collisions.collision)(label, previousLabel)) && arc2.data.id !== "thinner-slices";
                    if (showLabel) {
                      previousLabel = label;
                    }
                    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_react.Fragment, { children: [
                      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                        import_PieChartSlice.PieChartSlice,
                        {
                          sliceId: `slice-${arc2.data.id}`,
                          arc: arc2,
                          data: arc2.data,
                          path: pie.path,
                          isHighlighted
                        }
                      ),
                      showLabel && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                        import_PieChartLabel.PieChartLabel,
                        {
                          x: label.textX,
                          y: label.textY,
                          text: label.text,
                          isHighlighted
                        }
                      )
                    ] }, `slice-${arc2.data.id}`);
                  });
                }
              }
            )
          ]
        }
      ) })
    ] });
  }
);
