// @dynatrace/strato-components/esm/charts/xy-chart/utils/build-series-action-item.js
import { safeGet } from "../../core/utils/safe-get.js";
import { isCategoricalAxis } from "../types/guards.js";
import {
  BAR_VARIANT_SINGLE,
  BAR_VARIANT_STACK
} from "../types/xy-chart.js";
function baselineDirectionToAccessor(baseline) {
  return baseline ? `${baseline}0` : void 0;
}
function getBaselineCategoryInfo(datapoint, baselineAccessor, metadata) {
  const { xAxis, yAxis } = metadata;
  const axis = baselineAccessor === "x0" ? xAxis : yAxis;
  const value = datapoint[baselineAccessor];
  if (!isCategoricalAxis(axis) || typeof value !== "string") {
    return {
      category: void 0,
      isBaselineCategorical: false
    };
  }
  return {
    category: value,
    isBaselineCategorical: true
  };
}
const barSeriesFilterStrategy = (_series, currentMetadata, _seriesToMetadata, haveSameGrowingAxis) => {
  return currentMetadata.variant === BAR_VARIANT_STACK && !haveSameGrowingAxis;
};
const areaSeriesFilterStrategy = (series, _currentMetadata, seriesToMetadata, haveSameGrowingAxis) => {
  if (!haveSameGrowingAxis) {
    return true;
  }
  const seriesMetadata = safeGet(seriesToMetadata, series);
  return seriesMetadata.variant === "band";
};
const areaSimpleActionItemStrategy = (metadata) => metadata.variant === "band";
const barSimpleActionItemStrategy = (metadata) => metadata.variant === BAR_VARIANT_SINGLE;
function buildSeriesActionItemInternal({
  datapoint,
  allSeries,
  seriesToMetadata,
  datapointMetadataMap,
  currentSeries,
  simpleActionItemStrategy,
  seriesFilterStrategy
}) {
  const currentMetadata = safeGet(seriesToMetadata, currentSeries);
  const { baseline } = safeGet(datapointMetadataMap, datapoint);
  const baselineAccessor = baselineDirectionToAccessor(baseline);
  if (!baselineAccessor || simpleActionItemStrategy(currentMetadata)) {
    return { datapoint, name: currentMetadata.name };
  }
  const { category, isBaselineCategorical } = getBaselineCategoryInfo(
    datapoint,
    baselineAccessor,
    currentMetadata
  );
  if (!isBaselineCategorical || category === void 0) {
    return { datapoint, name: currentMetadata.name };
  }
  const relatedDatapoints = [];
  const primaryAxisAccessor = baseline === "x" ? "xAxisId" : "yAxisId";
  const growingAxisAccessor = baseline === "x" ? "yAxisId" : "xAxisId";
  for (const series of allSeries) {
    const haveSamePrimaryAxis = series[primaryAxisAccessor] === currentSeries[primaryAxisAccessor];
    const haveSameGrowingAxis = series[growingAxisAccessor] === currentSeries[growingAxisAccessor];
    if (!haveSamePrimaryAxis || seriesFilterStrategy(
      series,
      currentMetadata,
      seriesToMetadata,
      haveSameGrowingAxis
    )) {
      continue;
    }
    for (const dp of series.data) {
      if (dp[baselineAccessor] === category) {
        relatedDatapoints.push(dp);
      }
    }
  }
  return {
    datapoint,
    category,
    datapoints: relatedDatapoints,
    name: currentMetadata.name
  };
}
function buildBarSeriesActionItem(params) {
  return buildSeriesActionItemInternal({
    ...params,
    simpleActionItemStrategy: barSimpleActionItemStrategy,
    seriesFilterStrategy: barSeriesFilterStrategy
  });
}
function buildAreaSeriesActionItem(params) {
  return buildSeriesActionItemInternal({
    ...params,
    simpleActionItemStrategy: areaSimpleActionItemStrategy,
    seriesFilterStrategy: areaSeriesFilterStrategy
  });
}
export {
  buildAreaSeriesActionItem,
  buildBarSeriesActionItem
};
