// @dynatrace/strato-components/esm/charts/xy-chart/utils/calculate-axes-domains.js
import {
  XYAxisTypeCategorical,
  XYAxisTypeLinear,
  XYAxisTypeLog,
  XYAxisTypeTime
} from "../types/axis.js";
function getMinMax(axisId, coordinate0, coordinate1, axisIdToScaleDomain) {
  const minMax = axisIdToScaleDomain.get(axisId) ?? {
    min: coordinate0,
    max: coordinate1
  };
  const minValues = [minMax.min, coordinate0, coordinate1].filter(
    (value) => value !== void 0
  );
  const maxValues = [minMax.max, coordinate0, coordinate1].filter(
    (value) => value !== void 0
  );
  minMax.min = Math.min(...minValues);
  minMax.max = Math.max(...maxValues);
  return minMax;
}
function getCategoricalDomain(axisId, v0, v1, axisIdToScaleDomain) {
  const categoricalSet = axisIdToScaleDomain.get(axisId) ?? /* @__PURE__ */ new Set();
  if (v0) {
    categoricalSet.add(v0);
  }
  if (v1) {
    categoricalSet.add(v1);
  }
  return categoricalSet;
}
const calculateAxesDomainsMap = {
  [XYAxisTypeLinear]: (args) => {
    return getMinMax(
      args.axisId,
      args.a0,
      args.a1,
      args.axisIdToScaleDomain
    );
  },
  [XYAxisTypeLog]: (args) => {
    return getMinMax(
      args.axisId,
      args.a0,
      args.a1,
      args.axisIdToScaleDomain
    );
  },
  [XYAxisTypeCategorical]: (args) => {
    return getCategoricalDomain(
      args.axisId,
      args.a0,
      args.a1,
      args.axisIdToScaleDomain
    );
  },
  [XYAxisTypeTime]: (args) => {
    return getMinMax(
      args.axisId,
      args.a0?.valueOf(),
      args.a1?.valueOf(),
      args.axisIdToScaleDomain
    );
  }
};
function calculateAxesDomains(xAxis, yAxis, datapoint, axisIdToScaleDomain, isSeriesVisible) {
  const xDomain = calculateAxesDomainsMap[xAxis.type]({
    axisId: xAxis.id,
    a0: datapoint.x0,
    a1: datapoint.x1,
    axisIdToScaleDomain
  });
  const yDomain = isSeriesVisible || yAxis.type === XYAxisTypeCategorical ? calculateAxesDomainsMap[yAxis.type]({
    axisId: yAxis.id,
    a0: datapoint.y0,
    a1: datapoint.y1,
    axisIdToScaleDomain
  }) : axisIdToScaleDomain.get(yAxis.id) ?? {};
  return { xDomain, yDomain };
}
export {
  calculateAxesDomains
};
