// @dynatrace/strato-components/esm/charts/xy-chart/utils/generate-scales-by-axis-id.js
import { scaleBand, scaleTime } from "d3-scale";
import {
  niceUnitDomain,
  shouldUseNiceUnitTicks
} from "./generate-nice-unit-ticks.js";
import { isXAxis } from "./is-axis.js";
import { toRelative } from "./value-representation.js";
import {
  CATEGORICAL_BARS_INNER_PADDING,
  CATEGORICAL_BARS_OUTER_PADDING
} from "../../core/constants.js";
import {
  getMinMax,
  getScale
} from "../../core/utils/core-scale.js";
import { resolveTimeMinMax } from "../../core/utils/get-min-max-for-scale.js";
import { safeGet } from "../../core/utils/safe-get.js";
import {
  isLogarithmicAxis,
  isNumericalAxis,
  isTimeAxis
} from "../types/guards.js";
function generateInitialNumericalScale(initialDomain, axis, isXCoordinate, width, height, minMaxDomain) {
  const scaleType = axis.type.split(".")[1];
  let min, max;
  const initialDomainWithRelativeFilter = axis.isRelative ? {
    min: toRelative(initialDomain.min, minMaxDomain),
    max: toRelative(initialDomain.max, minMaxDomain)
  } : initialDomain;
  if (axis.interactions.initial) {
    ({ min, max } = initialDomainWithRelativeFilter);
  } else {
    const minMaxByConfig = { min: axis.min, max: axis.max };
    const niceMin = isLogarithmicAxis(axis) ? false : axis.min === "auto";
    const niceMax = isLogarithmicAxis(axis) ? false : axis.max === "auto";
    const rawMinMax = getMinMax(
      scaleType,
      initialDomainWithRelativeFilter,
      minMaxByConfig,
      { niceMin: false, niceMax: false }
    );
    const axisFormatter = axis.formatter;
    const niceDomain = shouldUseNiceUnitTicks(axis) && (niceMin || niceMax) ? niceUnitDomain(
      [rawMinMax.min, rawMinMax.max],
      axisFormatter,
      niceMin,
      niceMax
    ) : null;
    ({ min, max } = niceDomain ?? getMinMax(scaleType, initialDomainWithRelativeFilter, minMaxByConfig, {
      niceMin,
      niceMax
    }));
  }
  const domain = axis.reversed ? [max, min] : [min, max];
  const range = isXCoordinate ? [0, width] : [height, 0];
  return getScale(scaleType, domain, range);
}
function generateNumericScale(initialScale, zoomedDomain, scaleType) {
  const range = initialScale.range();
  let domain = initialScale.domain();
  if (zoomedDomain) {
    const min = zoomedDomain.min;
    const max = zoomedDomain.max;
    domain = [min, max];
  }
  return getScale(scaleType, domain, range);
}
function generateTimeScale(initialScale, zoomedDomain) {
  const range = initialScale.range();
  let domain;
  if (zoomedDomain) {
    const min = zoomedDomain.min;
    const max = zoomedDomain.max;
    domain = [min, max];
  } else {
    domain = initialScale.domain();
  }
  return scaleTime().domain(domain).range(range);
}
function generateInitialTimeScale(initialDomain, axis, isXCoordinate, width, height) {
  let min, max;
  if (axis.interactions.initial) {
    ({ min, max } = initialDomain);
  } else {
    const minMaxByConfig = { min: axis.min, max: axis.max };
    ({ min, max } = resolveTimeMinMax(initialDomain, minMaxByConfig));
  }
  const domain = axis.reversed ? [max, min] : [min, max];
  const range = isXCoordinate ? [0, width] : [height, 0];
  const scale = scaleTime().domain(domain).range(range);
  return scale;
}
function generateCategoricalScale(categoricalValues, axis, isXCoordinate, width, height) {
  const range = isXCoordinate ? [0, width] : [0, height];
  const domain = axis.reversed ? Array.from(categoricalValues).reverse() : categoricalValues;
  return scaleBand(domain, range);
}
function generateScalesByAxisId(axes, axisIdToScaleDomain, axisIdToZoomedDomain, axisToDimension, width, height) {
  const axisIdToScales = /* @__PURE__ */ new Map();
  axes.forEach((axis) => {
    const isXCoordinate = isXAxis(axis);
    const minMaxDomain = safeGet(axisIdToScaleDomain, axis.id);
    if (isNumericalAxis(axis)) {
      const initialDomain = (axis.interactions.initial && {
        min: axis.interactions.initial[0],
        max: axis.interactions.initial[1]
      }) ?? minMaxDomain;
      const initialScale = generateInitialNumericalScale(
        initialDomain,
        axis,
        isXCoordinate,
        width,
        height,
        minMaxDomain
      );
      const zoomedDomain = axisIdToZoomedDomain?.get(axis.id);
      const currentScale = generateNumericScale(
        initialScale,
        zoomedDomain,
        axis.type.split(".")[1]
      );
      axisIdToScales.set(axis.id, { currentScale, initialScale });
    } else if (isTimeAxis(axis)) {
      const initialDomain = (axis.interactions.initial && {
        min: axis.interactions.initial[0],
        max: axis.interactions.initial[1]
      }) ?? safeGet(axisIdToScaleDomain, axis.id);
      const initialScale = generateInitialTimeScale(
        initialDomain,
        axis,
        isXCoordinate,
        width,
        height
      );
      const zoomedDomain = axisIdToZoomedDomain?.get(axis.id);
      const currentScale = generateTimeScale(
        initialScale,
        zoomedDomain
      );
      axisIdToScales.set(axis.id, { currentScale, initialScale });
    } else {
      const initialDomain = safeGet(
        axisIdToScaleDomain,
        axis.id
      );
      const scale = generateCategoricalScale(
        initialDomain,
        axis,
        isXCoordinate,
        width,
        height
      );
      const dimensions = axisToDimension.get(axis.id);
      axisIdToScales.set(axis.id, {
        initialScale: scale,
        currentScale: scale,
        dimensionScale: dimensions ? scaleBand(dimensions, [
          0,
          scale.copy().paddingInner(CATEGORICAL_BARS_INNER_PADDING).paddingOuter(CATEGORICAL_BARS_OUTER_PADDING).bandwidth()
        ]).paddingInner(CATEGORICAL_BARS_INNER_PADDING) : void 0,
        categoricalBarScale: scale.copy().paddingInner(CATEGORICAL_BARS_INNER_PADDING).paddingOuter(CATEGORICAL_BARS_OUTER_PADDING)
      });
    }
  });
  return axisIdToScales;
}
export {
  generateInitialTimeScale,
  generateScalesByAxisId
};
