// @dynatrace/strato-components/esm/charts/xy-chart/utils/get-ticks-formatted-by-config.js
import { adaptivelyReduceHorizontalTicks } from "../../core/utils/adaptively-reduce-horizontal-ticks.js";
import {
  getCategoricalTicks,
  getDateNumericalFormattedTicks,
  getFormattedTicks
} from "../../core/utils/formatter-by-ticks.js";
import { isCategorical, isLinear, isLogarithmic } from "../types/guards.js";
const MIN_TICK_COUNT = 2;
function getTicksFormattedByConfig(config, numTicks, scale, formatter, availableWidth) {
  const axisType = config?.type || "";
  if (isCategorical(axisType)) {
    return getCategoricalTicks(scale, formatter);
  }
  const getTicksForCount = isLogarithmic(axisType) ? (count) => getFormattedTicks(
    count,
    scale,
    void 0,
    formatter
  ) : (count) => getDateNumericalFormattedTicks(
    count,
    scale,
    isLinear(axisType) && !config?.isRelative ? config?.unit : void 0,
    formatter
  );
  const result = getTicksForCount(numTicks);
  if (availableWidth === void 0 || scale === void 0) {
    return result;
  }
  let initialResult = result;
  if (result.formattedTicks.length < MIN_TICK_COUNT) {
    const boosted = getTicksForCount(MIN_TICK_COUNT);
    if (boosted.formattedTicks.length >= MIN_TICK_COUNT) {
      initialResult = boosted;
    } else {
      return result;
    }
  }
  const continuousScale = scale;
  const rawDomain = typeof scale.domain === "function" ? scale.domain() : null;
  const domainBounds = rawDomain !== null && rawDomain.length >= 2 ? [+rawDomain[0], +rawDomain[rawDomain.length - 1]] : void 0;
  const reduced = adaptivelyReduceHorizontalTicks(
    isLogarithmic(axisType) ? createLogMajorTickThinningBuilder(initialResult, MIN_TICK_COUNT) : createHalvingHintCandidateBuilder(
      initialResult,
      getTicksForCount,
      MIN_TICK_COUNT
    ),
    continuousScale,
    { domainBounds }
  );
  return {
    ...reduced,
    numericTicks: reduced.rawTicks.map(Number)
  };
}
function createLogMajorTickThinningBuilder(initialResult, minTickCount) {
  const allTicks = initialResult.ticks;
  const majorIndices = initialResult.formattedTicks.reduce(
    (acc, label, i) => {
      if (label !== "") {
        acc.push(i);
      }
      return acc;
    },
    []
  );
  return (attempt) => {
    if (attempt === 0) {
      return {
        ...initialResult,
        rawTicks: allTicks
      };
    }
    const step = Math.pow(2, attempt);
    const selectedMajorIndices = majorIndices.filter(
      (_, i) => i === 0 || i === majorIndices.length - 1 || i % step === 0
    );
    if (selectedMajorIndices.length < minTickCount) {
      return null;
    }
    const selectedSet = new Set(selectedMajorIndices);
    const newFormattedTicks = initialResult.formattedTicks.map((label, i) => {
      if (label === "") {
        return "";
      }
      return selectedSet.has(i) ? label : "";
    });
    return {
      ...initialResult,
      formattedTicks: newFormattedTicks,
      rawTicks: allTicks
    };
  };
}
function createHalvingHintCandidateBuilder(initialResult, getTicksForCount, minTickCount) {
  let hint = initialResult.formattedTicks.length;
  return (attempt) => {
    if (attempt === 0) {
      return {
        ...initialResult,
        rawTicks: initialResult.ticks
      };
    }
    hint = Math.max(Math.floor(hint / 2), minTickCount);
    const next = getTicksForCount(hint);
    return { ...next, rawTicks: next.ticks };
  };
}
export {
  createHalvingHintCandidateBuilder,
  getTicksFormattedByConfig
};
