// @dynatrace/strato-components/esm/charts/xy-chart/utils/query-closest-quadtree.js
import { quadtree } from "d3-quadtree";
import { debounce } from "lodash-es";
import {
  getLinearAccessors
} from "./get-linear-accessors.js";
import { getCurrentScale } from "./get-scale.js";
import { relAccessor } from "./value-representation.js";
import { DEFAULT_DATA_POINT_RADIUS } from "../../core/constants.js";
import { safeGet } from "../../core/utils/safe-get.js";
function getAxisAccessors(series, metadata, xScale, yScale, axisIdToScaleDomain, datapointToMetadata) {
  const { xAxis, yAxis } = metadata;
  const xRelAccessor = relAccessor(xAxis, axisIdToScaleDomain);
  const yRelAccessor = relAccessor(yAxis, axisIdToScaleDomain);
  const { xAccessor, yAccessor } = getLinearAccessors(
    series.type,
    metadata,
    datapointToMetadata,
    xRelAccessor,
    yRelAccessor
  );
  return [xAccessor(xScale), yAccessor(yScale)];
}
function filterDatapointsInRange(points, xAxisAccessor, yAxisAccessor, bounds, datapointToMetadata) {
  const { xMin, xMax, yMin, yMax } = bounds;
  return points.filter((point) => {
    if (safeGet(datapointToMetadata, point).isGap) {
      return false;
    }
    const xValue = xAxisAccessor(point);
    const yValue = yAxisAccessor(point);
    if (xValue < xMin || xValue > xMax) {
      return false;
    }
    if (yValue < yMin || yValue > yMax) {
      return false;
    }
    return true;
  });
}
function buildSingleQuadtree(xAxisAccessor, yAxisAccessor, filtered) {
  return quadtree().x((d) => xAxisAccessor(d)).y((d) => yAxisAccessor(d)).addAll(filtered);
}
function buildQuadtrees(seriesWithDatapoints, axisIdToScales, datapointToMetadata, seriesToMetadata, axisIdToScaleDomain) {
  const seriesToTrees = /* @__PURE__ */ new Map();
  for (const series of seriesWithDatapoints) {
    const xScale = getCurrentScale(axisIdToScales, series.xAxisId);
    const yScale = getCurrentScale(axisIdToScales, series.yAxisId);
    const metadata = safeGet(seriesToMetadata, series);
    const accessors = getAxisAccessors(
      series,
      metadata,
      xScale,
      yScale,
      axisIdToScaleDomain,
      datapointToMetadata
    );
    const [xAxisAccessor, yAxisAccessor] = accessors;
    const [xMin, xMax] = xScale.range();
    const [yMin, yMax] = yScale.range();
    const filtered = filterDatapointsInRange(
      series.data,
      xAxisAccessor,
      yAxisAccessor,
      {
        xMin: Math.min(xMin, xMax),
        xMax: Math.max(xMin, xMax),
        yMin: Math.min(yMin, yMax),
        yMax: Math.max(yMin, yMax)
      },
      datapointToMetadata
    );
    const tree = buildSingleQuadtree(xAxisAccessor, yAxisAccessor, filtered);
    seriesToTrees.set(series.id, tree);
  }
  return seriesToTrees;
}
const DELAY = 300;
const debounceMap = /* @__PURE__ */ new Map();
function getDebouncedQuadtrees(key) {
  let debounced = debounceMap.get(key);
  if (!debounced) {
    debounced = debounce(
      (series, scales, metadata, seriesToMetadata, axisIdToScaleDomain, onDone) => onDone(
        buildQuadtrees(
          series,
          scales,
          metadata,
          seriesToMetadata,
          axisIdToScaleDomain
        )
      ),
      DELAY
    );
    debounceMap.set(key, debounced);
  }
  return debounced;
}
function cancelDebouncedQuadtrees(key) {
  debounceMap.get(key)?.cancel?.();
}
function distanceBetweenTwoPoints(x0, x1, y0, y1) {
  return Math.sqrt(Math.pow(x0 - x1, 2) + Math.pow(y0 - y1, 2));
}
function getDatapointAxisValues(datapoint, series, metadata, xScale, yScale, axisIdToScaleDomain, datapointToMetadata) {
  const accessors = getAxisAccessors(
    series,
    metadata,
    xScale,
    yScale,
    axisIdToScaleDomain,
    datapointToMetadata
  );
  return [accessors[0](datapoint), accessors[1](datapoint)];
}
function findClosestDatapointInSeries(series, tree, position, metadata, queryContext) {
  const datapoint = tree.find(position.x, position.y);
  if (!datapoint) {
    return void 0;
  }
  const [xValue, yValue] = getDatapointAxisValues(
    datapoint,
    series,
    metadata,
    queryContext.xScale,
    queryContext.yScale,
    queryContext.axisIdToScaleDomain,
    queryContext.datapointToMetadata
  );
  const distance = distanceBetweenTwoPoints(
    position.x,
    xValue,
    position.y,
    yValue
  );
  return {
    type: series.type,
    distance,
    datapoint,
    series
  };
}
function buildQueryClosest(seriesIdToQuadtree, axisIdToScales, seriesToMetadata, axisIdToScaleDomain, datapointToMetadata) {
  function query(arrayOfSeries, position) {
    let best;
    for (const series of arrayOfSeries) {
      const tree = seriesIdToQuadtree.get(series.id);
      if (!tree) {
        continue;
      }
      const xScale = getCurrentScale(axisIdToScales, series.xAxisId);
      const yScale = getCurrentScale(axisIdToScales, series.yAxisId);
      const metadata = safeGet(seriesToMetadata, series);
      const closest = findClosestDatapointInSeries(
        series,
        tree,
        position,
        metadata,
        {
          xScale,
          yScale,
          axisIdToScaleDomain,
          datapointToMetadata
        }
      );
      if (!closest) {
        continue;
      }
      if (!best || closest.distance < best.distance) {
        best = closest;
      }
      if (best && best.distance <= DEFAULT_DATA_POINT_RADIUS) {
        break;
      }
    }
    return best;
  }
  return query;
}
export {
  buildQuadtrees,
  buildQueryClosest,
  cancelDebouncedQuadtrees,
  distanceBetweenTwoPoints,
  getDebouncedQuadtrees
};
