// @dynatrace/strato-components/esm/charts/xy-chart/utils/query-hovered.js
import { getCurrentScale, getDimensionScale } from "./get-scale.js";
import { isRectDatapoint, isRectSeries } from "./type-guards.js";
import { relAccessor } from "./value-representation.js";
import { safeGet } from "../../core/utils/safe-get.js";
import {
  xAccessor as barXAccessor,
  yAccessor as barYAccessor,
  widthAccessor as barWidthAccessor,
  heightAccessor as barHeightAccessor
} from "../components/displays/bar-accessors.js";
import {
  xAccessor as rectXAccessor,
  yAccessor as rectYAccessor,
  widthAccessor as rectWidthAccessor,
  heightAccessor as rectHeightAccessor
} from "../components/displays/rect-accessors.js";
import {
  RECT_VARIANT
} from "../types/xy-chart-internals.js";
function isInPixelRange(position, start, span) {
  return position >= start && position <= start + span;
}
function getAccessors(series, axisIdToScales, datapointMetadataMap, xRelAccessor, yRelAccessor) {
  const { type, xAxisId, yAxisId } = series;
  const xScale = getCurrentScale(axisIdToScales, xAxisId, type);
  const yScale = getCurrentScale(axisIdToScales, yAxisId, type);
  if (type === RECT_VARIANT) {
    return {
      xAccessor: rectXAccessor(xScale, xRelAccessor),
      yAccessor: rectYAccessor(yScale, yRelAccessor),
      widthAccessor: rectWidthAccessor(xScale, xRelAccessor),
      heightAccessor: rectHeightAccessor(yScale, yRelAccessor)
    };
  }
  const dimensionXScale = getDimensionScale(axisIdToScales, xAxisId);
  const dimensionYScale = getDimensionScale(axisIdToScales, yAxisId);
  return {
    xAccessor: barXAccessor(
      xScale,
      datapointMetadataMap,
      dimensionXScale,
      series.id,
      xRelAccessor
    ),
    yAccessor: barYAccessor(
      yScale,
      datapointMetadataMap,
      dimensionYScale,
      series.id,
      yRelAccessor
    ),
    widthAccessor: barWidthAccessor(
      xScale,
      datapointMetadataMap,
      dimensionXScale,
      xRelAccessor
    ),
    heightAccessor: barHeightAccessor(
      yScale,
      datapointMetadataMap,
      dimensionYScale,
      yRelAccessor
    )
  };
}
const getPixelDimensionsForDatapoint = (datapoint, genAccessors) => {
  if (isRectDatapoint(datapoint)) {
    const accessors2 = genAccessors;
    return {
      x: accessors2.xAccessor(datapoint),
      y: accessors2.yAccessor(datapoint),
      width: accessors2.widthAccessor(datapoint),
      height: accessors2.heightAccessor(datapoint)
    };
  }
  const accessors = genAccessors;
  return {
    x: accessors.xAccessor(datapoint),
    y: accessors.yAccessor(datapoint),
    width: accessors.widthAccessor(datapoint),
    height: accessors.heightAccessor(datapoint)
  };
};
function getHoveredDatapoint(position, series, axisIdToScales, datapointMetadataMap, seriesToMetadata, axisIdToScaleDomain) {
  const data = isRectSeries(series) ? series.filteredData : series.data;
  if (position.x < 0 || position.y < 0) {
    return void 0;
  }
  const { xAxis, yAxis } = safeGet(seriesToMetadata, series);
  const xRelAccessor = relAccessor(xAxis, axisIdToScaleDomain);
  const yRelAccessor = relAccessor(yAxis, axisIdToScaleDomain);
  const accessors = getAccessors(
    series,
    axisIdToScales,
    datapointMetadataMap,
    xRelAccessor,
    yRelAccessor
  );
  for (const datapoint of data) {
    if (safeGet(datapointMetadataMap, datapoint).isGap) {
      continue;
    }
    const { x, y, width, height } = getPixelDimensionsForDatapoint(
      datapoint,
      accessors
    );
    if (isInPixelRange(position.x, x, width) && isInPixelRange(position.y, y, height)) {
      return datapoint;
    }
  }
  return void 0;
}
function buildQueryHovered(axisIdToScales, datapointMetadataMap, seriesToMetadata, axisIdToScaleDomain) {
  function query(arrayOfSeries, position) {
    for (let i = arrayOfSeries.length - 1; i >= 0; i--) {
      const series = arrayOfSeries[i];
      const datapoint = getHoveredDatapoint(
        position,
        series,
        axisIdToScales,
        datapointMetadataMap,
        seriesToMetadata,
        axisIdToScaleDomain
      );
      if (!datapoint) {
        continue;
      }
      return {
        type: series.type,
        distance: 0,
        datapoint,
        series
      };
    }
    return void 0;
  }
  return query;
}
export {
  buildQueryHovered,
  getHoveredDatapoint,
  getPixelDimensionsForDatapoint
};
