// @dynatrace/strato-components/esm/charts/xy-chart/utils/query-get-closest-point.js
import { isEmpty } from "lodash-es";
import {
  getLinearAccessors
} from "./get-linear-accessors.js";
import { getCurrentScale } from "./get-scale.js";
import { distanceBetweenTwoPoints } from "./query-closest-quadtree.js";
import { relAccessor } from "./value-representation.js";
import { DEFAULT_DATA_POINT_RADIUS } from "../../core/constants.js";
import { isBandScale } from "../../core/utils/detect-scale.js";
import { safeGet } from "../../core/utils/safe-get.js";
function getDistance(datapoint, xScale, yScale, relativePosition, xAccessor, yAccessor) {
  const xPixel = xAccessor(xScale)(datapoint);
  const yPixel = yAccessor(yScale)(datapoint);
  return distanceBetweenTwoPoints(
    relativePosition.x,
    xPixel,
    relativePosition.y,
    yPixel
  );
}
function calculateScalesBounds(xScale, yScale) {
  const [minWidth, maxWidth] = xScale.range();
  const yScaleRange = yScale.range();
  const [minHeight, maxHeight] = isBandScale(yScale) ? [yScaleRange[0], yScaleRange[1]] : [yScaleRange[1], yScaleRange[0]];
  return { minWidth, maxWidth, minHeight, maxHeight };
}
function getAccessorsForSeries(seriesType, seriesToMetadata, currentSeries, datapointToMetadata, xRelAccessor, yRelAccessor) {
  const metadata = safeGet(seriesToMetadata, currentSeries);
  return getLinearAccessors(
    seriesType,
    metadata,
    datapointToMetadata,
    xRelAccessor,
    yRelAccessor
  );
}
function isDatapointOutOfBounds(xPixel, yPixel, bounds) {
  const { minWidth, maxWidth, minHeight, maxHeight } = bounds;
  return xPixel < minWidth || xPixel > maxWidth || yPixel < minHeight || yPixel > maxHeight;
}
function findClosestInSeriesData(data, xScale, yScale, relativePosition, accessors, bounds, currentAccumulator) {
  let bestInSeries = currentAccumulator;
  let datapointIndex = 0;
  while (datapointIndex < data.length) {
    const datapoint = data[datapointIndex];
    const xPixel = accessors.xAccessor(xScale)(datapoint);
    const yPixel = accessors.yAccessor(yScale)(datapoint);
    if (isDatapointOutOfBounds(xPixel, yPixel, bounds)) {
      datapointIndex += 1;
      continue;
    }
    const distance = getDistance(
      datapoint,
      xScale,
      yScale,
      relativePosition,
      accessors.xAccessor,
      accessors.yAccessor
    );
    if (distance < bestInSeries.distance) {
      bestInSeries = {
        distance,
        datapoint,
        series: currentAccumulator.series
      };
    }
    if (distance <= DEFAULT_DATA_POINT_RADIUS) {
      return bestInSeries;
    }
    datapointIndex += 1;
  }
  return bestInSeries.distance < currentAccumulator.distance ? bestInSeries : void 0;
}
function getClosestDatapoint(relativePosition, series, axisIdToScales, seriesToMetadata, datapointToMetadata, axisIdToScaleDomain) {
  if (!series.length) {
    return void 0;
  }
  const initialAccumulator = {
    distance: Infinity,
    datapoint: {},
    series: {}
  };
  let bestMatch = initialAccumulator;
  let seriesIndex = series.length - 1;
  while (seriesIndex >= 0) {
    const currentSeries = series[seriesIndex];
    const { data, xAxisId, yAxisId, type } = currentSeries;
    const xScale = getCurrentScale(axisIdToScales, xAxisId);
    const yScale = getCurrentScale(axisIdToScales, yAxisId);
    const seriesMetadata = safeGet(seriesToMetadata, currentSeries);
    const xRelAccessor = axisIdToScaleDomain ? relAccessor(seriesMetadata.xAxis, axisIdToScaleDomain) : void 0;
    const yRelAccessor = axisIdToScaleDomain ? relAccessor(seriesMetadata.yAxis, axisIdToScaleDomain) : void 0;
    const accessors = getAccessorsForSeries(
      type,
      seriesToMetadata,
      currentSeries,
      datapointToMetadata,
      xRelAccessor,
      yRelAccessor
    );
    const bounds = calculateScalesBounds(xScale, yScale);
    const seriesAccumulator = {
      ...initialAccumulator,
      series: currentSeries
    };
    const updatedAccumulator = findClosestInSeriesData(
      data,
      xScale,
      yScale,
      relativePosition,
      accessors,
      bounds,
      seriesAccumulator
    );
    if (updatedAccumulator) {
      bestMatch = updatedAccumulator;
      if (updatedAccumulator.distance <= DEFAULT_DATA_POINT_RADIUS) {
        break;
      }
    }
    seriesIndex -= 1;
  }
  if (isEmpty(bestMatch.datapoint) || isEmpty(bestMatch.series)) {
    return void 0;
  }
  return {
    datapoint: bestMatch.datapoint,
    series: bestMatch.series,
    distance: bestMatch.distance
  };
}
function buildQueryGetClosestPoint(axisIdToScales, seriesToMetadata, datapointToMetadata, axisIdToScaleDomain) {
  function query(arrayOfSeries, position) {
    if (isEmpty(arrayOfSeries)) {
      return void 0;
    }
    const closest = getClosestDatapoint(
      position,
      arrayOfSeries,
      axisIdToScales,
      seriesToMetadata,
      datapointToMetadata,
      axisIdToScaleDomain
    );
    if (!closest) {
      return void 0;
    }
    return {
      type: closest.series.type,
      distance: closest?.distance,
      datapoint: closest?.datapoint,
      series: closest?.series
    };
  }
  return query;
}
export {
  buildQueryGetClosestPoint,
  getClosestDatapoint,
  getDistance
};
