// @dynatrace/strato-components/esm/charts/xy-chart/utils/type-guards.js
import {
  AREA_VARIANT,
  BAR_VARIANT,
  DOT_VARIANT,
  LINE_VARIANT,
  RECT_VARIANT
} from "../types/xy-chart-internals.js";
function isNumerical(args) {
  return typeof args.v0 === "number";
}
function isCategorical(args) {
  return typeof args.v0 === "string";
}
function isTime(args) {
  return args.v0 instanceof Date;
}
function isNumericalHoveringPoint(args) {
  return isNumerical(args);
}
function isTimeHoveringPoint(args) {
  return isTime(args);
}
function isCategoricalHoveringPoint(args) {
  return isCategorical(args);
}
function isNumericalDotScale(args) {
  return isNumerical(args);
}
function isTimeDotScale(args) {
  return isTime(args);
}
function isCategoricalDotScale(args) {
  return isCategorical(args);
}
function isNumericalRectScale(args) {
  return isNumerical(args);
}
function isTimeRectScale(args) {
  return isTime(args);
}
function isCategoricalRectScale(args) {
  return isCategorical(args);
}
function isNumericalLineScale(args) {
  return isNumerical(args);
}
function isTimeLineScale(args) {
  return isTime(args);
}
function isCategoricalLineScale(args) {
  return isCategorical(args);
}
function isNumericalBarScale(args) {
  return typeof args.v0 === "number" || typeof args.v1 === "number";
}
function isTimeBarScale(args) {
  return args.v0 instanceof Date || args.v1 instanceof Date;
}
function isCategoricalBarScale(args) {
  return typeof args.v0 === "string" || typeof args.v1 === "string";
}
function isNumericalAreaScale(args) {
  return typeof args.v0 === "number" || typeof args.v1 === "number";
}
function isTimeAreaScale(args) {
  return args.v0 instanceof Date || args.v1 instanceof Date;
}
function isCategoricalAreaScale(args) {
  return typeof args.v0 === "string" || typeof args.v1 === "string";
}
function isNumericalToolTip(args) {
  return isNumerical(args);
}
function isTimeTooltip(args) {
  return isTime(args);
}
function isCategoricalTooltip(args) {
  return isCategorical(args);
}
function isRectSeries(series) {
  return series.type === RECT_VARIANT;
}
function isDotSeries(series) {
  return series.type === DOT_VARIANT;
}
function isLineSeries(series) {
  return series.type === LINE_VARIANT;
}
function isBarSeries(series) {
  return series.type === BAR_VARIANT;
}
function isAreaSeries(series) {
  return series.type === AREA_VARIANT;
}
function isRectTooltipContext(context) {
  return context.series.type === "rect" && !!context.metadata;
}
function isBarTooltipContext(context) {
  return context.series.type === "bar" && !!context.metadata;
}
function isAreaTooltipContext(context) {
  return context.series.type === "area" && !!context.metadata;
}
function isDotTooltipContent(content) {
  return content?.series?.type === DOT_VARIANT;
}
function isLineTooltipContent(content) {
  return content?.series?.type === LINE_VARIANT;
}
function isAreaTooltipContent(content) {
  return content?.series?.type === AREA_VARIANT;
}
function isRectDatapoint(datapoint) {
  return "value" in datapoint;
}
const isStringOrValidNumber = (value) => {
  return typeof value === "string" || typeof value === "number" && !Number.isNaN(value);
};
function isValidName(value) {
  return value !== null && value !== void 0 && !Number.isNaN(value);
}
function isHorizontalAreaAccessors(accessors) {
  return accessors.x.length === 4;
}
function isVerticalAreaAccessors(accessors) {
  return accessors.y.length === 4;
}
export {
  isAreaSeries,
  isAreaTooltipContent,
  isAreaTooltipContext,
  isBarSeries,
  isBarTooltipContext,
  isCategoricalAreaScale,
  isCategoricalBarScale,
  isCategoricalDotScale,
  isCategoricalHoveringPoint,
  isCategoricalLineScale,
  isCategoricalRectScale,
  isCategoricalTooltip,
  isDotSeries,
  isDotTooltipContent,
  isHorizontalAreaAccessors,
  isLineSeries,
  isLineTooltipContent,
  isNumericalAreaScale,
  isNumericalBarScale,
  isNumericalDotScale,
  isNumericalHoveringPoint,
  isNumericalLineScale,
  isNumericalRectScale,
  isNumericalToolTip,
  isRectDatapoint,
  isRectSeries,
  isRectTooltipContext,
  isStringOrValidNumber,
  isTimeAreaScale,
  isTimeBarScale,
  isTimeDotScale,
  isTimeHoveringPoint,
  isTimeLineScale,
  isTimeRectScale,
  isTimeTooltip,
  isValidName,
  isVerticalAreaAccessors
};
