// @dynatrace/strato-components/esm/charts/xy-chart/utils/xy-chart-configuration.js
import { isEmpty, merge } from "lodash-es";
import { interactionsFor } from "./axis-interactions.js";
import { doesNothing } from "./defaults.js";
import { isXAxis } from "./is-axis.js";
import { TRUNCATION_MODE } from "../../core/constants.js";
import { isValidColorRule } from "../../core/slots/ColorRule.js";
import { isCustomColorPalette } from "../../core/types/color-palette.js";
import { isColoredRangeArray } from "../../core/types/colored-range.js";
import { isValidColorPalette } from "../../core/utils/is-valid-color-palette.js";
import { isValidLegendPosition } from "../../core/utils/is-valid-legend-position.js";
import { isValidToolbarPlacement } from "../../core/utils/is-valid-toolbar-placement.js";
import { buildThresholds } from "../components/thresholds/builder.js";
import {
  DEFAULT_AXIS_CONFIG,
  defaultConfig,
  defaultLegend,
  defaultToolbar
} from "../constants/defaults.js";
import {
  DEFAULT_CUSTOM_TOOLTIP_CONFIG
} from "../context/XYChartCustomTooltip.context.js";
import { XYAxisMapper } from "../types/axis.js";
import {
  NUMERICAL_SCALE_TYPE
} from "../types/xy-chart.js";
function getLegendConfiguration(slotsLegend, contextLegend) {
  const legendConfig = merge(
    {},
    defaultLegend,
    contextLegend,
    slotsLegend
  );
  legendConfig.position = isValidLegendPosition(legendConfig.position) ? legendConfig.position : defaultLegend.position;
  return legendConfig;
}
function getToolbarConfiguration(slotsToolbar, contextToolbar) {
  const toolbarConfig = merge({}, defaultToolbar, contextToolbar, slotsToolbar);
  toolbarConfig.placement = isValidToolbarPlacement(toolbarConfig.placement) ? toolbarConfig.placement : defaultToolbar.placement;
  return toolbarConfig;
}
function getAxesWithSeries(axisSlots, series) {
  let axesConfig = [];
  const matchingIds = /* @__PURE__ */ new Set();
  axisSlots.forEach((axis) => {
    const isAlreadyDefined = matchingIds.has(axis.id);
    if (isAlreadyDefined) {
      return;
    }
    const hasMatchingSeriesId = series.find((series2) => {
      return isXAxis(axis) ? series2.xAxisId === axis.id : series2.yAxisId === axis.id;
    });
    if (hasMatchingSeriesId) {
      matchingIds.add(axis.id);
      axesConfig = [...axesConfig, axis];
    }
  });
  return axesConfig;
}
function getColorPaletteConfiguration(contextColorPalette, propsColorPalette) {
  const colorPalette = propsColorPalette ?? contextColorPalette ?? void 0;
  if (!colorPalette) {
    return void 0;
  }
  return isColoredRangeArray(colorPalette) || isCustomColorPalette(colorPalette) || isValidColorPalette(colorPalette) ? colorPalette : void 0;
}
function getTruncationModeConfiguration(contextTruncationMode, propsTruncationMode) {
  return propsTruncationMode ?? contextTruncationMode ?? TRUNCATION_MODE.Middle;
}
function getXYChartAnnotationsConfiguration(slotsAnnotations, configAnnotations) {
  if (isEmpty(slotsAnnotations) && !configAnnotations) {
    return void 0;
  }
  return merge({}, configAnnotations, slotsAnnotations);
}
function getCustomTooltipConfiguration(slotsCustomTooltip, contextCustomTooltip) {
  if (slotsCustomTooltip !== void 0) {
    return slotsCustomTooltip;
  }
  return {
    ...DEFAULT_CUSTOM_TOOLTIP_CONFIG,
    hidden: contextCustomTooltip?.hidden ?? DEFAULT_CUSTOM_TOOLTIP_CONFIG.hidden
  };
}
const createXYChartConfiguration = (slots, contextConfig, rootProps) => {
  const colorPalette = getColorPaletteConfiguration(
    contextConfig.colorPalette,
    rootProps?.colorPalette
  );
  const truncationMode = getTruncationModeConfiguration(
    contextConfig.truncationMode,
    rootProps?.truncationMode
  );
  const legend = getLegendConfiguration(slots.legend, contextConfig.legend);
  const xAxes = getAxes(slots.xAxes, contextConfig.xAxes);
  const yAxes = getAxes(slots.yAxes, contextConfig.yAxes);
  const toolbar = getToolbarConfiguration(slots.toolbar, contextConfig.toolbar);
  const annotations = getXYChartAnnotationsConfiguration(
    slots.annotations,
    contextConfig.annotations
  );
  const select = slots.select || void 0;
  const thresholdIndicators = [
    ...slots.thresholds,
    ...contextConfig.thresholdIndicators || []
  ];
  const colorRules = (contextConfig.colorRules || []).filter(isValidColorRule).concat(slots.colorRules || []);
  const tooltip = getCustomTooltipConfiguration(
    slots.tooltip,
    contextConfig.tooltip
  );
  return {
    colorPalette,
    legend,
    toolbar,
    truncationMode,
    xAxes,
    yAxes,
    thresholdIndicators,
    annotations,
    select,
    colorRules,
    tooltip,
    customTooltip: {}
  };
};
function getAxes(slotAxes, contextAxes) {
  const configContextAxes = contextAxes ?? [];
  const ids = /* @__PURE__ */ new Set();
  const axesConfig = [...slotAxes, ...configContextAxes];
  return axesConfig.reduce((acc, axis) => {
    if (!ids.has(axis.id)) {
      acc.push({ ...DEFAULT_AXIS_CONFIG, ...axis });
      ids.add(axis.id);
    }
    return acc;
  }, []);
}
const deserializeConfig = (serializedConfig) => {
  try {
    return merge({}, defaultConfig, JSON.parse(serializedConfig));
  } catch {
    return defaultConfig;
  }
};
const buildInternalConfig = (config, slots) => {
  const yAxes = config.yAxes.map((axis) => ({
    ...axis,
    coordinate: "y",
    tickLabelLayout: "horizontal",
    type: XYAxisMapper[axis.type],
    interactions: { zoom: false, pan: false, infinite: false },
    onZoomChange: doesNothing,
    isRelative: axis.type === NUMERICAL_SCALE_TYPE && axis.valueRepresentation === "relative"
  }));
  const xAxes = config.xAxes.map((axis) => ({
    ...axis,
    coordinate: "x",
    tickLabelLayout: axis.tickLabelLayout ?? "horizontal",
    type: XYAxisMapper[axis.type],
    interactions: interactionsFor(axis),
    onZoomChange: axis.onZoomChange ?? doesNothing,
    isRelative: axis.type === NUMERICAL_SCALE_TYPE && axis.valueRepresentation === "relative"
  }));
  const axes = getAxesWithSeries(
    [...yAxes, ...xAxes],
    slots.series
  );
  const thresholds = buildThresholds(yAxes, xAxes, config.thresholdIndicators);
  return { series: slots.series, axes, thresholds };
};
export {
  buildInternalConfig,
  createXYChartConfiguration,
  deserializeConfig,
  getAxes,
  getTruncationModeConfiguration,
  getXYChartAnnotationsConfiguration
};
