// @dynatrace/strato-components/esm/charts/timeseries/components/annotations-timeseries/TimeseriesChartAnnotationsArea.js
import { jsx } from "react/jsx-runtime";
import { ANNOTATIONS_PADDING_BOTTOM } from "./constants.js";
import { useBehavioralTrackingProps } from "../../../../core/hooks/useBehavioralTrackingProps.js";
import { AnnotationsTimeDisplay } from "../../../core/components/annotations/components/displays/AnnotationsTimeDisplay.js";
import { AnnotationsScrollWrapper } from "../../../core/components/annotations/components/overlays/AnnotationsScrollWrapper.js";
import {
  DEFAULT_INTEGRATION_TEXT_OVERFLOW,
  DEFAULT_SHOW_LABELS,
  SCROLLBAR_THRESHOLD
} from "../../../core/components/annotations/constants.js";
import { ChartGrid } from "../../../core/components/chart-grid/ChartGridRoot.js";
import { useGridPlotSize } from "../../../core/components/chart-grid/hooks/useGridPlotSize.js";
import { useSetState } from "../../hooks/store/selectors.js";
import { useAnnotationsConfig } from "../../hooks/useAnnotationsConfig.js";
import { useXScales } from "../../hooks/useXScale.js";
import { AREAS } from "../../types/timeseries-grid-area-rules.js";
const COMPONENT_NAME = "AnnotationsChart";
const TimeseriesChartAnnotationsArea = () => {
  const config = useAnnotationsConfig();
  const {
    tracks,
    showLabels = DEFAULT_SHOW_LABELS,
    annotationsActions,
    tooltip
  } = config;
  const [behaviorTrackingProps] = useBehavioralTrackingProps(config);
  const setState = useSetState();
  const { currentScale } = useXScales();
  const scale = currentScale;
  const { width } = useGridPlotSize();
  const handleUpdateIndicators = (event) => {
    setState((prev) => ({
      ...prev,
      annotations: {
        ...prev.annotations,
        ...event
      }
    }));
  };
  return /* @__PURE__ */ jsx(
    ChartGrid.GridArea,
    {
      paddingBottom: ANNOTATIONS_PADDING_BOTTOM,
      gridArea: AREAS.annotations,
      children: ({ availableWidth, availableHeight }) => {
        const labelOffset = availableWidth - width;
        return /* @__PURE__ */ jsx(
          AnnotationsScrollWrapper,
          {
            height: availableHeight,
            width: availableWidth + SCROLLBAR_THRESHOLD,
            children: /* @__PURE__ */ jsx(
              AnnotationsTimeDisplay,
              {
                tracks,
                width: availableWidth,
                scale,
                showLabels: showLabels ?? DEFAULT_SHOW_LABELS,
                labelsOffset: labelOffset,
                annotationsActions: (annotation) => annotation && annotationsActions?.(annotation),
                markerGroupHover: handleUpdateIndicators,
                textOverflow: DEFAULT_INTEGRATION_TEXT_OVERFLOW,
                tooltipConfig: tooltip,
                ...behaviorTrackingProps,
                "data-dt-component": COMPONENT_NAME
              }
            )
          }
        );
      }
    }
  );
};
export {
  TimeseriesChartAnnotationsArea
};
