// @dynatrace/strato-components/content/chip-group/ChipGroup.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ChipGroup_exports = {};
__export(ChipGroup_exports, {
  ChipGroup: () => ChipGroup
});
module.exports = __toCommonJS(ChipGroup_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_ChipGroup_css = require("./ChipGroup.sty.js");
var import_ChipGroupContext = require("./contexts/ChipGroupContext.js");
var import_InternalChipGroupContext = require("./contexts/InternalChipGroupContext.js");
var import_Control = require("./Control.js");
var import_HiddenCount = require("./HiddenCount.js");
var import_chip_group_state_reducer = require("./state/chip-group-state-reducer.js");
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useContainerSize = require("../../core/hooks/useContainerSize.js");
var import_useControllableState = require("../../core/hooks/useControllableState.js");
var import_useId = require("../../core/hooks/useId.js");
var import_useMergeRefs = require("../../core/hooks/useMergeRefs.js");
var import_useSafeForwardProps = require("../../core/hooks/useSafeForwardProps.js");
var import_Flex = require("../../layouts/flex/Flex.js");
const COMPONENT_NAME = "ChipGroup";
const RootChipGroup = (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const {
      children,
      expanded: expandedProp,
      defaultExpanded,
      disabled,
      size,
      onExpandedChange,
      maxVisibleChips = "auto",
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
    const ariaLabelingProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(restProps);
    const forwardProps = (0, import_useSafeForwardProps.useSafeForwardProps)(restProps, ariaLabelingProps);
    const uniqueId = (0, import_useId.useId)();
    const [state, dispatch] = (0, import_react.useReducer)(import_chip_group_state_reducer.chipGroupReducer, {
      chips: {}
    });
    const containerRef = (0, import_react.useRef)(null);
    const rootRef = (0, import_useMergeRefs.useMergeRefs)([
      containerRef,
      forwardedRef
    ]);
    const { width: containerWidth } = (0, import_useContainerSize.useContainerSize)(containerRef);
    const [expanded, setExpanded] = (0, import_useControllableState.useControllableState)({
      value: expandedProp,
      defaultValue: defaultExpanded,
      onChange: onExpandedChange
    });
    const [allChipsFit, setAllChipsFit] = (0, import_react.useState)(false);
    const [hiddenChips, setHiddenChips] = (0, import_react.useState)([]);
    const [visibleChips, setVisibleChips] = (0, import_react.useState)([]);
    (0, import_react.useEffect)(() => {
      const chips = Object.entries(state.chips);
      const chipKeys = chips.map(([uniqueChipKey]) => uniqueChipKey);
      if (chips.length === 0) {
        setVisibleChips([]);
        return;
      }
      if (chips.some(([, chip]) => !chip.width)) {
        return;
      }
      if (expanded) {
        setHiddenChips([]);
        setVisibleChips(chipKeys);
        if (!containerRef.current?.firstChild) {
          setAllChipsFit(true);
          return;
        }
        let wrap = false;
        let child = containerRef.current.firstChild;
        const rafId = requestAnimationFrame(() => {
          do {
            if (child.classList.contains("chip") && child.offsetTop > 0) {
              wrap = true;
            }
            child = child.nextSibling;
          } while (!wrap && child);
          setAllChipsFit(!wrap);
          setTimeout(() => {
            const controlButton = containerRef.current?.querySelector(
              `button[aria-controls="${uniqueId}"]`
            );
            if (controlButton) {
              controlButton.focus();
            }
          }, 0);
        });
        return () => {
          cancelAnimationFrame(rafId);
        };
      }
      const spacing = size === "condensed" ? 6 : 8;
      let totalChipsWidth = 0;
      let setHidden = true;
      let _hiddenChipsIndex = chips.length;
      for (const [index, [, value]] of Array.from(chips).entries()) {
        totalChipsWidth += value.width || 0;
        const chipFits = index === 0 || Math.floor(totalChipsWidth + index * spacing) <= Math.round(containerWidth);
        const chipAndControlFits = index === 0 || totalChipsWidth + (index + 1) * spacing + (state.control?.width || 0) <= containerWidth;
        const visible = index === chips.length - 1 && (state.control?.show ?? "auto") === "auto" ? chipFits : chipAndControlFits;
        if (!visible && setHidden) {
          setHidden = false;
          _hiddenChipsIndex = index;
        }
      }
      const effectiveHiddenChipsIndex = maxVisibleChips !== "auto" && maxVisibleChips >= 0 ? Math.min(_hiddenChipsIndex, maxVisibleChips) : _hiddenChipsIndex;
      const allFit = effectiveHiddenChipsIndex === chips.length;
      setAllChipsFit(allFit);
      setHiddenChips(chipKeys.slice(effectiveHiddenChipsIndex));
      setVisibleChips(chipKeys.slice(0, effectiveHiddenChipsIndex));
    }, [containerWidth, expanded, size, state, maxVisibleChips, uniqueId]);
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_ChipGroupContext.ChipGroupContext.Provider,
      {
        value: {
          visibleKeys: visibleChips,
          hiddenKeys: hiddenChips
        },
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_InternalChipGroupContext.InternalChipGroupContext.Provider,
          {
            value: {
              chipGroupId: uniqueId,
              disabled,
              size,
              expanded,
              allChipsFit,
              dispatch,
              setExpanded
            },
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_Flex.Flex,
              {
                ref: rootRef,
                id: uniqueId,
                className: (0, import_clsx.default)(
                  (0, import_ChipGroup_css.chipGroupCSS)({ size, expanded }),
                  consumerClassName
                ),
                style: consumerStyle,
                "data-testid": dataTestId,
                "data-dtrum-mask": dataDtrumMask,
                "data-dtrum-allow": dataDtrumAllow,
                ...ariaLabelingProps,
                ...behavioralTrackingProps,
                ...forwardProps,
                "data-dt-component": COMPONENT_NAME,
                children
              }
            )
          }
        )
      }
    );
  }
);
RootChipGroup.displayName = COMPONENT_NAME;
const ChipGroup = Object.assign(RootChipGroup, {
  Control: import_Control.Control,
  HiddenCount: import_HiddenCount.HiddenCount
});
