// @dynatrace/strato-components/content/chip-group/Control.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Control_exports = {};
__export(Control_exports, {
  Control: () => Control
});
module.exports = __toCommonJS(Control_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_clsx = __toESM(require("clsx"));
var import_react = require("react");
var import_react_intl = require("react-intl");
var styles = __toESM(require("./Control.sty.js"));
var import_useChipGroupContext = require("./hooks/useChipGroupContext.js");
var import_useInternalChipGroupContext = require("./hooks/useInternalChipGroupContext.js");
var import_chip_group_state_reducer = require("./state/chip-group-state-reducer.js");
var import_useContainerSize = require("../../core/hooks/useContainerSize.js");
var import_useFocusRing = require("../../core/styles/useFocusRing.js");
var import_ProgressCircle = require("../progress/ProgressCircle.js");
const Control = (props) => {
  const {
    display,
    count,
    loading = false,
    onClick,
    children,
    className: consumerClassName,
    style: consumerStyle,
    "data-testid": dataTestId
  } = props;
  const intl = (0, import_react_intl.useIntl)();
  const { focusProps, focusClassName, isFocusVisible } = (0, import_useFocusRing.useFocusRing)({
    color: "neutral"
  });
  const { hiddenKeys, visibleKeys } = (0, import_useChipGroupContext.useChipGroupContext)();
  const internalChipGroupContext = (0, import_useInternalChipGroupContext.useInternalChipGroupContext)();
  const { chipGroupId, expanded, allChipsFit, setExpanded, dispatch } = internalChipGroupContext || {};
  const controlRef = (0, import_react.useRef)(null);
  const { width: controlWidth } = (0, import_useContainerSize.useContainerSize)(controlRef);
  const width = Math.ceil(controlWidth);
  const label = (0, import_react.useMemo)(() => {
    const moreOrLess = expanded && display !== "more" ? "less" : "more";
    const hiddenChipsCount = count === true && hiddenKeys.length > 0 ? hiddenKeys.length : count;
    const showCount = typeof hiddenChipsCount === "number" && moreOrLess === "more";
    return intl.formatMessage(
      {
        defaultMessage: "Show {count} {moreOrLess}",
        id: "ARdI7d8cJxHnUayP",
        description: "Label for the expand / collapse button."
      },
      {
        moreOrLess: expanded && display !== "more" ? "less" : "more",
        count: showCount ? hiddenChipsCount : void 0
      }
    );
  }, [count, display, expanded, hiddenKeys, intl]);
  (0, import_react.useEffect)(() => {
    dispatch?.({
      type: import_chip_group_state_reducer.ChipGroupActions.ADDCONTROL,
      payload: { width }
    });
    return () => {
      dispatch?.({
        type: import_chip_group_state_reducer.ChipGroupActions.REMOVECONTROL,
        payload: {}
      });
    };
  }, []);
  (0, import_react.useEffect)(() => {
    dispatch?.({
      type: import_chip_group_state_reducer.ChipGroupActions.UPDATECONTROL,
      payload: {
        width: width === 0 ? void 0 : width,
        show: display !== void 0 ? "always" : "auto"
      }
    });
  }, [dispatch, display, width]);
  const shouldHide = display === void 0 && (allChipsFit || visibleKeys.length === 0) && !loading;
  if (shouldHide) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    "button",
    {
      "data-testid": dataTestId,
      ref: controlRef,
      onClick: (event) => {
        onClick?.(event);
        setExpanded?.((prev) => display === "more" ? true : !prev);
      },
      "aria-controls": chipGroupId,
      "aria-expanded": expanded,
      className: (0, import_clsx.default)(
        styles.expandButton({ isFocusVisible }),
        focusClassName,
        consumerClassName
      ),
      style: consumerStyle,
      ...focusProps,
      children: [
        loading && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_ProgressCircle.ProgressCircle,
          {
            size: "small",
            "aria-label": intl.formatMessage({
              defaultMessage: "Loading",
              id: "cycz+bYG+VH7Of4e",
              description: "Label for the loading state."
            })
          }
        ),
        children ?? label
      ]
    }
  );
};
