// @dynatrace/strato-components/tables/DataTable/tanstack-table-core-override/src/features/ColumnFiltering.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ColumnFiltering_exports = {};
__export(ColumnFiltering_exports, {
  ColumnFiltering: () => ColumnFiltering,
  shouldAutoRemoveFilter: () => shouldAutoRemoveFilter
});
module.exports = __toCommonJS(ColumnFiltering_exports);
var import__ = require("..");
var import_filterFns = require("../filterFns");
var import_utils = require("../utils");
const ColumnFiltering = {
  getDefaultColumnDef: () => {
    return {
      filterFn: "auto"
    };
  },
  getInitialState: (state) => {
    return {
      columnFilters: [],
      ...state
    };
  },
  getDefaultOptions: (table) => {
    return {
      onColumnFiltersChange: (0, import_utils.makeStateUpdater)("columnFilters", table),
      filterFromLeafRows: false,
      maxLeafRowFilterDepth: 100
    };
  },
  createColumn: (column, table) => {
    column.getAutoFilterFn = () => {
      const firstRow = table.getCoreRowModel().flatRows[0];
      const value = firstRow?.getValue(column.id);
      if (typeof value === "string") {
        return import_filterFns.filterFns.includesString;
      }
      if (typeof value === "number") {
        return import_filterFns.filterFns.inNumberRange;
      }
      if (typeof value === "boolean") {
        return import_filterFns.filterFns.equals;
      }
      if (value !== null && typeof value === "object") {
        return import_filterFns.filterFns.equals;
      }
      if (Array.isArray(value)) {
        return import_filterFns.filterFns.arrIncludes;
      }
      return import_filterFns.filterFns.weakEquals;
    };
    column.getFilterFn = () => {
      return (0, import_utils.isFunction)(column.columnDef.filterFn) ? column.columnDef.filterFn : column.columnDef.filterFn === "auto" ? column.getAutoFilterFn() : (
        // @ts-ignore
        table.options.filterFns?.[column.columnDef.filterFn] ?? import_filterFns.filterFns[column.columnDef.filterFn]
      );
    };
    column.getCanFilter = () => {
      return (column.columnDef.enableColumnFilter ?? true) && (table.options.enableColumnFilters ?? true) && (table.options.enableFilters ?? true) && !!column.accessorFn;
    };
    column.getIsFiltered = () => column.getFilterIndex() > -1;
    column.getFilterValue = () => table.getState().columnFilters?.find((d) => d.id === column.id)?.value;
    column.getFilterIndex = () => table.getState().columnFilters?.findIndex((d) => d.id === column.id) ?? -1;
    column.setFilterValue = (value) => {
      table.setColumnFilters((old) => {
        const filterFn = column.getFilterFn();
        const previousFilter = old?.find((d) => d.id === column.id);
        const newFilter = (0, import_utils.functionalUpdate)(
          value,
          previousFilter ? previousFilter.value : void 0
        );
        if (shouldAutoRemoveFilter(filterFn, newFilter, column)) {
          return old?.filter((d) => d.id !== column.id) ?? [];
        }
        const newFilterObj = { id: column.id, value: newFilter };
        if (previousFilter) {
          return old?.map((d) => {
            if (d.id === column.id) {
              return newFilterObj;
            }
            return d;
          }) ?? [];
        }
        if (old?.length) {
          return [...old, newFilterObj];
        }
        return [newFilterObj];
      });
    };
  },
  createTable: (table) => {
    table.setColumnFilters = (updater) => {
      const leafColumns = table.getAllLeafColumns();
      const updateFn = (old) => {
        return (0, import_utils.functionalUpdate)(updater, old)?.filter((filter) => {
          const column = leafColumns.find((d) => d.id === filter.id);
          if (column) {
            const filterFn = column.getFilterFn();
            if (shouldAutoRemoveFilter(filterFn, filter.value, column)) {
              return false;
            }
          }
          return true;
        });
      };
      table.options.onColumnFiltersChange?.(updateFn);
    };
    table.resetColumnFilters = (defaultState) => {
      table.setColumnFilters(
        defaultState ? [] : table.initialState?.columnFilters ?? []
      );
    };
    table.getPreFilteredRowModel = () => table.getCoreRowModel();
    table.getFilteredRowModel = () => {
      if (!table._getFilteredRowModel && table.options.getFilteredRowModel) {
        table._getFilteredRowModel = table.options.getFilteredRowModel(table);
      }
      if (table.options.manualFiltering || !table._getFilteredRowModel) {
        return table.getPreFilteredRowModel();
      }
      return table._getFilteredRowModel();
    };
    Object.defineProperties((0, import__.getRowProto)(table), {
      columnFilters: {
        get() {
          return this._columnFilters ??= {};
        },
        set(value) {
          this._columnFilters = value;
        },
        enumerable: true
      },
      columnFiltersMeta: {
        get() {
          return this._columnFiltersMeta ??= {};
        },
        set(value) {
          this._columnFiltersMeta = value;
        },
        enumerable: true
      }
    });
  }
};
function shouldAutoRemoveFilter(filterFn, value, column) {
  return (filterFn && filterFn.autoRemove ? filterFn.autoRemove(value, column) : false) || typeof value === "undefined" || typeof value === "string" && !value;
}
