// @dynatrace/strato-components/tables/DataTable/tanstack-table-core-override/src/features/ColumnVisibility.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ColumnVisibility_exports = {};
__export(ColumnVisibility_exports, {
  ColumnVisibility: () => ColumnVisibility,
  _getVisibleLeafColumns: () => _getVisibleLeafColumns
});
module.exports = __toCommonJS(ColumnVisibility_exports);
var import__ = require("..");
var import_utils = require("../utils");
const ColumnVisibility = {
  getInitialState: (state) => {
    return {
      columnVisibility: {},
      ...state
    };
  },
  getDefaultOptions: (table) => {
    return {
      onColumnVisibilityChange: (0, import_utils.makeStateUpdater)("columnVisibility", table)
    };
  },
  createColumn: (column, table) => {
    column.toggleVisibility = (value) => {
      if (column.getCanHide()) {
        table.setColumnVisibility((old) => ({
          ...old,
          [column.id]: value ?? !column.getIsVisible()
        }));
      }
    };
    column.getIsVisible = () => {
      const childColumns = column.columns;
      return (childColumns.length ? childColumns.some((c) => c.getIsVisible()) : table.getState().columnVisibility?.[column.id]) ?? true;
    };
    column.getCanHide = () => {
      return (column.columnDef.enableHiding ?? true) && (table.options.enableHiding ?? true);
    };
    column.getToggleVisibilityHandler = () => {
      return (e) => {
        column.toggleVisibility?.(
          e.target.checked
        );
      };
    };
  },
  createTable: (table) => {
    const makeVisibleColumnsMethod = (key, getColumns) => {
      return (0, import_utils.memo)(
        () => [
          getColumns(),
          getColumns().filter((d) => d.getIsVisible()).map((d) => d.id).join("_")
        ],
        (columns) => {
          return columns.filter((d) => d.getIsVisible?.());
        },
        (0, import_utils.getMemoOptions)(table.options, "debugColumns", key)
      );
    };
    table.getVisibleFlatColumns = makeVisibleColumnsMethod(
      "getVisibleFlatColumns",
      () => table.getAllFlatColumns()
    );
    table.getVisibleLeafColumns = makeVisibleColumnsMethod(
      "getVisibleLeafColumns",
      () => table.getAllLeafColumns()
    );
    table.getLeftVisibleLeafColumns = makeVisibleColumnsMethod(
      "getLeftVisibleLeafColumns",
      () => table.getLeftLeafColumns()
    );
    table.getRightVisibleLeafColumns = makeVisibleColumnsMethod(
      "getRightVisibleLeafColumns",
      () => table.getRightLeafColumns()
    );
    table.getCenterVisibleLeafColumns = makeVisibleColumnsMethod(
      "getCenterVisibleLeafColumns",
      () => table.getCenterLeafColumns()
    );
    table.setColumnVisibility = (updater) => table.options.onColumnVisibilityChange?.(updater);
    table.resetColumnVisibility = (defaultState) => {
      table.setColumnVisibility(
        defaultState ? {} : table.initialState.columnVisibility ?? {}
      );
    };
    table.toggleAllColumnsVisible = (value) => {
      value = value ?? !table.getIsAllColumnsVisible();
      table.setColumnVisibility(
        table.getAllLeafColumns().reduce(
          (obj, column) => ({
            ...obj,
            [column.id]: !value ? !column.getCanHide?.() : value
          }),
          {}
        )
      );
    };
    table.getIsAllColumnsVisible = () => !table.getAllLeafColumns().some((column) => !column.getIsVisible?.());
    table.getIsSomeColumnsVisible = () => table.getAllLeafColumns().some((column) => column.getIsVisible?.());
    table.getToggleAllColumnsVisibilityHandler = () => {
      return (e) => {
        table.toggleAllColumnsVisible(
          e.target?.checked
        );
      };
    };
    Object.assign((0, import__.getRowProto)(table), {
      _getAllVisibleCells: (0, import_utils.memo)(
        function() {
          return [this.getAllCells(), table.getState().columnVisibility];
        },
        (cells) => {
          return cells.filter((cell) => cell.column.getIsVisible());
        },
        (0, import_utils.getMemoOptions)(table.options, "debugRows", "_getAllVisibleCells")
      ),
      getVisibleCells: (0, import_utils.memo)(
        function() {
          return [
            this.getLeftVisibleCells(),
            this.getCenterVisibleCells(),
            this.getRightVisibleCells()
          ];
        },
        (left, center, right) => [...left, ...center, ...right],
        (0, import_utils.getMemoOptions)(table.options, "debugRows", "getVisibleCells")
      )
    });
  }
};
function _getVisibleLeafColumns(table, position) {
  return !position ? table.getVisibleLeafColumns() : position === "center" ? table.getCenterVisibleLeafColumns() : position === "left" ? table.getLeftVisibleLeafColumns() : table.getRightVisibleLeafColumns();
}
