// @dynatrace/strato-components/tables/DataTable/tanstack-table-core-override/src/features/ColumnOrdering.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ColumnOrdering_exports = {};
__export(ColumnOrdering_exports, {
  ColumnOrdering: () => ColumnOrdering
});
module.exports = __toCommonJS(ColumnOrdering_exports);
var import_utils = require("../utils");
var import_ColumnGrouping = require("./ColumnGrouping");
var import__ = require("..");
const ColumnOrdering = {
  getInitialState: (state) => {
    return {
      columnOrder: [],
      ...state
    };
  },
  getDefaultOptions: (table) => {
    return {
      onColumnOrderChange: (0, import_utils.makeStateUpdater)("columnOrder", table)
    };
  },
  createColumn: (column, table) => {
    column.getIndex = (0, import_utils.memo)(
      (position) => [(0, import__._getVisibleLeafColumns)(table, position)],
      (columns) => columns.findIndex((d) => d.id === column.id),
      (0, import_utils.getMemoOptions)(table.options, "debugColumns", "getIndex")
    );
    column.getIsFirstColumn = (position) => {
      const columns = (0, import__._getVisibleLeafColumns)(table, position);
      return columns[0]?.id === column.id;
    };
    column.getIsLastColumn = (position) => {
      const columns = (0, import__._getVisibleLeafColumns)(table, position);
      return columns[columns.length - 1]?.id === column.id;
    };
  },
  createTable: (table) => {
    table.setColumnOrder = (updater) => table.options.onColumnOrderChange?.(updater);
    table.resetColumnOrder = (defaultState) => {
      table.setColumnOrder(
        defaultState ? [] : table.initialState.columnOrder ?? []
      );
    };
    table._getOrderColumnsFn = (0, import_utils.memo)(
      () => [
        table.getState().columnOrder,
        table.getState().grouping,
        table.options.groupedColumnMode
      ],
      (columnOrder, grouping, groupedColumnMode) => (columns) => {
        let orderedColumns = [];
        if (!columnOrder?.length) {
          orderedColumns = columns;
        } else {
          const columnOrderCopy = [...columnOrder];
          const columnsCopy = [...columns];
          while (columnsCopy.length && columnOrderCopy.length) {
            const targetColumnId = columnOrderCopy.shift();
            const foundIndex = columnsCopy.findIndex(
              (d) => d.id === targetColumnId
            );
            if (foundIndex > -1) {
              orderedColumns.push(columnsCopy.splice(foundIndex, 1)[0]);
            }
          }
          orderedColumns = [...orderedColumns, ...columnsCopy];
        }
        return (0, import_ColumnGrouping.orderColumns)(orderedColumns, grouping, groupedColumnMode);
      },
      (0, import_utils.getMemoOptions)(table.options, "debugTable", "_getOrderColumnsFn")
    );
  }
};
