// @dynatrace/strato-components/tables/DataTable/tanstack-table-core-override/src/features/ColumnSizing.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ColumnSizing_exports = {};
__export(ColumnSizing_exports, {
  ColumnSizing: () => ColumnSizing,
  defaultColumnSizing: () => defaultColumnSizing,
  passiveEventSupported: () => passiveEventSupported
});
module.exports = __toCommonJS(ColumnSizing_exports);
var import__ = require("..");
var import_utils = require("../utils");
const defaultColumnSizing = {
  size: 150,
  minSize: 20,
  maxSize: Number.MAX_SAFE_INTEGER
};
const getDefaultColumnSizingInfoState = () => ({
  startOffset: null,
  startSize: null,
  deltaOffset: null,
  deltaPercentage: null,
  isResizingColumn: false,
  columnSizingStart: []
});
const ColumnSizing = {
  getDefaultColumnDef: () => {
    return defaultColumnSizing;
  },
  getInitialState: (state) => {
    return {
      columnSizing: {},
      columnSizingInfo: getDefaultColumnSizingInfoState(),
      ...state
    };
  },
  getDefaultOptions: (table) => {
    return {
      columnResizeMode: "onEnd",
      columnResizeDirection: "ltr",
      onColumnSizingChange: (0, import_utils.makeStateUpdater)("columnSizing", table),
      onColumnSizingInfoChange: (0, import_utils.makeStateUpdater)("columnSizingInfo", table)
    };
  },
  createColumn: (column, table) => {
    column.getSize = () => {
      const columnSize = table.getState().columnSizing[column.id];
      return Math.min(
        Math.max(
          column.columnDef.minSize ?? defaultColumnSizing.minSize,
          columnSize ?? column.columnDef.size ?? defaultColumnSizing.size
        ),
        column.columnDef.maxSize ?? defaultColumnSizing.maxSize
      );
    };
    column.getStart = (0, import_utils.memo)(
      (position) => [
        position,
        (0, import__._getVisibleLeafColumns)(table, position),
        table.getState().columnSizing
      ],
      (position, columns) => columns.slice(0, column.getIndex(position)).reduce((sum, column2) => sum + column2.getSize(), 0),
      (0, import_utils.getMemoOptions)(table.options, "debugColumns", "getStart")
    );
    column.getAfter = (0, import_utils.memo)(
      (position) => [
        position,
        (0, import__._getVisibleLeafColumns)(table, position),
        table.getState().columnSizing
      ],
      (position, columns) => columns.slice(column.getIndex(position) + 1).reduce((sum, column2) => sum + column2.getSize(), 0),
      (0, import_utils.getMemoOptions)(table.options, "debugColumns", "getAfter")
    );
    column.resetSize = () => {
      table.setColumnSizing(({ [column.id]: _, ...rest }) => {
        return rest;
      });
    };
    column.getCanResize = () => {
      return (column.columnDef.enableResizing ?? true) && (table.options.enableColumnResizing ?? true);
    };
    column.getIsResizing = () => {
      return table.getState().columnSizingInfo.isResizingColumn === column.id;
    };
  },
  createHeader: (header, table) => {
    header.getSize = () => {
      let sum = 0;
      const recurse = (header2) => {
        if (header2.subHeaders.length) {
          header2.subHeaders.forEach(recurse);
        } else {
          sum += header2.column.getSize() ?? 0;
        }
      };
      recurse(header);
      return sum;
    };
    header.getStart = () => {
      if (header.index > 0) {
        const prevSiblingHeader = header.headerGroup.headers[header.index - 1];
        return prevSiblingHeader.getStart() + prevSiblingHeader.getSize();
      }
      return 0;
    };
    header.getResizeHandler = (_contextDocument) => {
      const column = table.getColumn(header.column.id);
      const canResize = column?.getCanResize();
      return (e) => {
        if (!column || !canResize) {
          return;
        }
        e.persist?.();
        if (isTouchStartEvent(e)) {
          if (e.touches && e.touches.length > 1) {
            return;
          }
        }
        const startSize = header.getSize();
        const columnSizingStart = header ? header.getLeafHeaders().map((d) => [d.column.id, d.column.getSize()]) : [[column.id, column.getSize()]];
        const clientX = isTouchStartEvent(e) ? Math.round(e.touches[0].clientX) : e.clientX;
        const newColumnSizing = {};
        const updateOffset = (eventType, clientXPos) => {
          if (typeof clientXPos !== "number") {
            return;
          }
          table.setColumnSizingInfo((old) => {
            const deltaDirection = table.options.columnResizeDirection === "rtl" ? -1 : 1;
            const deltaOffset = (clientXPos - (old?.startOffset ?? 0)) * deltaDirection;
            const deltaPercentage = Math.max(
              deltaOffset / (old?.startSize ?? 0),
              -0.999999
            );
            old.columnSizingStart.forEach(([columnId, headerSize]) => {
              newColumnSizing[columnId] = Math.round(
                Math.max(headerSize + headerSize * deltaPercentage, 0) * 100
              ) / 100;
            });
            return {
              ...old,
              deltaOffset,
              deltaPercentage
            };
          });
          if (table.options.columnResizeMode === "onChange" || eventType === "end") {
            table.setColumnSizing((old) => ({
              ...old,
              ...newColumnSizing
            }));
          }
        };
        const onMove = (clientXPos) => updateOffset("move", clientXPos);
        const onEnd = (clientXPos) => {
          updateOffset("end", clientXPos);
          table.setColumnSizingInfo((old) => ({
            ...old,
            isResizingColumn: false,
            startOffset: null,
            startSize: null,
            deltaOffset: null,
            deltaPercentage: null,
            columnSizingStart: []
          }));
        };
        const contextDocument = _contextDocument || typeof document !== "undefined" ? document : null;
        const mouseEvents = {
          moveHandler: (e2) => onMove(e2.clientX),
          upHandler: (e2) => {
            contextDocument?.removeEventListener(
              "mousemove",
              mouseEvents.moveHandler
            );
            contextDocument?.removeEventListener(
              "mouseup",
              mouseEvents.upHandler
            );
            onEnd(e2.clientX);
          }
        };
        const touchEvents = {
          moveHandler: (e2) => {
            if (e2.cancelable) {
              e2.preventDefault();
              e2.stopPropagation();
            }
            onMove(e2.touches[0].clientX);
            return false;
          },
          upHandler: (e2) => {
            contextDocument?.removeEventListener(
              "touchmove",
              touchEvents.moveHandler
            );
            contextDocument?.removeEventListener(
              "touchend",
              touchEvents.upHandler
            );
            if (e2.cancelable) {
              e2.preventDefault();
              e2.stopPropagation();
            }
            onEnd(e2.touches[0]?.clientX);
          }
        };
        const passiveIfSupported = passiveEventSupported() ? { passive: false } : false;
        if (isTouchStartEvent(e)) {
          contextDocument?.addEventListener(
            "touchmove",
            touchEvents.moveHandler,
            passiveIfSupported
          );
          contextDocument?.addEventListener(
            "touchend",
            touchEvents.upHandler,
            passiveIfSupported
          );
        } else {
          contextDocument?.addEventListener(
            "mousemove",
            mouseEvents.moveHandler,
            passiveIfSupported
          );
          contextDocument?.addEventListener(
            "mouseup",
            mouseEvents.upHandler,
            passiveIfSupported
          );
        }
        table.setColumnSizingInfo((old) => ({
          ...old,
          startOffset: clientX,
          startSize,
          deltaOffset: 0,
          deltaPercentage: 0,
          columnSizingStart,
          isResizingColumn: column.id
        }));
      };
    };
  },
  createTable: (table) => {
    table.setColumnSizing = (updater) => table.options.onColumnSizingChange?.(updater);
    table.setColumnSizingInfo = (updater) => table.options.onColumnSizingInfoChange?.(updater);
    table.resetColumnSizing = (defaultState) => {
      table.setColumnSizing(
        defaultState ? {} : table.initialState.columnSizing ?? {}
      );
    };
    table.resetHeaderSizeInfo = (defaultState) => {
      table.setColumnSizingInfo(
        defaultState ? getDefaultColumnSizingInfoState() : table.initialState.columnSizingInfo ?? getDefaultColumnSizingInfoState()
      );
    };
    table.getTotalSize = () => table.getHeaderGroups()[0]?.headers.reduce((sum, header) => {
      return sum + header.getSize();
    }, 0) ?? 0;
    table.getLeftTotalSize = () => table.getLeftHeaderGroups()[0]?.headers.reduce((sum, header) => {
      return sum + header.getSize();
    }, 0) ?? 0;
    table.getCenterTotalSize = () => table.getCenterHeaderGroups()[0]?.headers.reduce((sum, header) => {
      return sum + header.getSize();
    }, 0) ?? 0;
    table.getRightTotalSize = () => table.getRightHeaderGroups()[0]?.headers.reduce((sum, header) => {
      return sum + header.getSize();
    }, 0) ?? 0;
  }
};
let passiveSupported = null;
function passiveEventSupported() {
  if (typeof passiveSupported === "boolean") return passiveSupported;
  let supported = false;
  try {
    const options = {
      get passive() {
        supported = true;
        return false;
      }
    };
    const noop = () => {
    };
    window.addEventListener("test", noop, options);
    window.removeEventListener("test", noop);
  } catch (err) {
    supported = false;
  }
  passiveSupported = supported;
  return passiveSupported;
}
function isTouchStartEvent(e) {
  return e.type === "touchstart";
}
