// @dynatrace/strato-components/tables/DataTable/tanstack-table-core-override/src/features/ColumnGrouping.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ColumnGrouping_exports = {};
__export(ColumnGrouping_exports, {
  ColumnGrouping: () => ColumnGrouping,
  orderColumns: () => orderColumns
});
module.exports = __toCommonJS(ColumnGrouping_exports);
var import__ = require("..");
var import_aggregationFns = require("../aggregationFns");
var import_utils = require("../utils");
const ColumnGrouping = {
  getDefaultColumnDef: () => {
    return {
      aggregatedCell: (props) => props.getValue()?.toString?.() ?? null,
      aggregationFn: "auto"
    };
  },
  getInitialState: (state) => {
    return {
      grouping: [],
      ...state
    };
  },
  getDefaultOptions: (table) => {
    return {
      onGroupingChange: (0, import_utils.makeStateUpdater)("grouping", table),
      groupedColumnMode: "reorder"
    };
  },
  createColumn: (column, table) => {
    column.toggleGrouping = () => {
      table.setGrouping((old) => {
        if (old?.includes(column.id)) {
          return old.filter((d) => d !== column.id);
        }
        return [...old ?? [], column.id];
      });
    };
    column.getCanGroup = () => {
      return (column.columnDef.enableGrouping ?? true) && (table.options.enableGrouping ?? true) && (!!column.accessorFn || !!column.columnDef.getGroupingValue);
    };
    column.getIsGrouped = () => {
      return table.getState().grouping?.includes(column.id);
    };
    column.getGroupedIndex = () => table.getState().grouping?.indexOf(column.id);
    column.getToggleGroupingHandler = () => {
      const canGroup = column.getCanGroup();
      return () => {
        if (!canGroup) return;
        column.toggleGrouping();
      };
    };
    column.getAutoAggregationFn = () => {
      const firstRow = table.getCoreRowModel().flatRows[0];
      const value = firstRow?.getValue(column.id);
      if (typeof value === "number") {
        return import_aggregationFns.aggregationFns.sum;
      }
      if (Object.prototype.toString.call(value) === "[object Date]") {
        return import_aggregationFns.aggregationFns.extent;
      }
    };
    column.getAggregationFn = () => {
      if (!column) {
        throw new Error();
      }
      return (0, import_utils.isFunction)(column.columnDef.aggregationFn) ? column.columnDef.aggregationFn : column.columnDef.aggregationFn === "auto" ? column.getAutoAggregationFn() : table.options.aggregationFns?.[column.columnDef.aggregationFn] ?? import_aggregationFns.aggregationFns[column.columnDef.aggregationFn];
    };
  },
  createTable: (table) => {
    table.setGrouping = (updater) => table.options.onGroupingChange?.(updater);
    table.resetGrouping = (defaultState) => {
      table.setGrouping(defaultState ? [] : table.initialState?.grouping ?? []);
    };
    table.getPreGroupedRowModel = () => table.getFilteredRowModel();
    table.getGroupedRowModel = () => {
      if (!table._getGroupedRowModel && table.options.getGroupedRowModel) {
        table._getGroupedRowModel = table.options.getGroupedRowModel(table);
      }
      if (table.options.manualGrouping || !table._getGroupedRowModel) {
        return table.getPreGroupedRowModel();
      }
      return table._getGroupedRowModel();
    };
    Object.defineProperty((0, import__.getRowProto)(table), "_groupingValuesCache", {
      get() {
        return this.__groupingValuesCache ??= {};
      },
      enumerable: true
    });
    Object.assign((0, import__.getRowProto)(table), {
      getIsGrouped() {
        return !!this.groupingColumnId;
      },
      getGroupingValue(columnId) {
        if (this._groupingValuesCache.hasOwnProperty(columnId)) {
          return this._groupingValuesCache[columnId];
        }
        const column = table.getColumn(columnId);
        if (!column?.columnDef.getGroupingValue) {
          return this.getValue(columnId);
        }
        this._groupingValuesCache[columnId] = column.columnDef.getGroupingValue(
          this.original
        );
        return this._groupingValuesCache[columnId];
      }
    });
  },
  createCell: (cell, column, row, table) => {
    const getRenderValue = () => cell.getValue() ?? table.options.renderFallbackValue;
    cell.getIsGrouped = () => column.getIsGrouped() && column.id === row.groupingColumnId;
    cell.getIsPlaceholder = () => !cell.getIsGrouped() && column.getIsGrouped();
    cell.getIsAggregated = () => !cell.getIsGrouped() && !cell.getIsPlaceholder() && !!row.subRows?.length;
  }
};
function orderColumns(leafColumns, grouping, groupedColumnMode) {
  if (!grouping?.length || !groupedColumnMode) {
    return leafColumns;
  }
  const nonGroupingColumns = leafColumns.filter(
    (col) => !grouping.includes(col.id)
  );
  if (groupedColumnMode === "remove") {
    return nonGroupingColumns;
  }
  const groupingColumns = grouping.map((g) => leafColumns.find((col) => col.id === g)).filter(Boolean);
  return [...groupingColumns, ...nonGroupingColumns];
}
