// @dynatrace/strato-components/tables/DataTable/tanstack-table-core-override/src/features/ColumnPinning.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ColumnPinning_exports = {};
__export(ColumnPinning_exports, {
  ColumnPinning: () => ColumnPinning
});
module.exports = __toCommonJS(ColumnPinning_exports);
var import__ = require("..");
var import_utils = require("../utils");
const getDefaultColumnPinningState = () => ({
  left: [],
  right: []
});
const ColumnPinning = {
  getInitialState: (state) => {
    return {
      columnPinning: getDefaultColumnPinningState(),
      ...state
    };
  },
  getDefaultOptions: (table) => {
    return {
      onColumnPinningChange: (0, import_utils.makeStateUpdater)("columnPinning", table)
    };
  },
  createColumn: (column, table) => {
    column.pin = (position) => {
      const columnIds = column.getLeafColumns().map((d) => d.id).filter(Boolean);
      table.setColumnPinning((old) => {
        if (position === "right") {
          return {
            left: (old?.left ?? []).filter((d) => !columnIds?.includes(d)),
            right: [
              ...(old?.right ?? []).filter((d) => !columnIds?.includes(d)),
              ...columnIds
            ]
          };
        }
        if (position === "left") {
          return {
            left: [
              ...(old?.left ?? []).filter((d) => !columnIds?.includes(d)),
              ...columnIds
            ],
            right: (old?.right ?? []).filter((d) => !columnIds?.includes(d))
          };
        }
        return {
          left: (old?.left ?? []).filter((d) => !columnIds?.includes(d)),
          right: (old?.right ?? []).filter((d) => !columnIds?.includes(d))
        };
      });
    };
    column.getCanPin = () => {
      const leafColumns = column.getLeafColumns();
      return leafColumns.some(
        (d) => (d.columnDef.enablePinning ?? true) && (table.options.enableColumnPinning ?? table.options.enablePinning ?? true)
      );
    };
    column.getIsPinned = () => {
      const leafColumnIds = column.getLeafColumns().map((d) => d.id);
      const { left, right } = table.getState().columnPinning;
      const isLeft = leafColumnIds.some((d) => left?.includes(d));
      const isRight = leafColumnIds.some((d) => right?.includes(d));
      return isLeft ? "left" : isRight ? "right" : false;
    };
    column.getPinnedIndex = () => {
      const position = column.getIsPinned();
      return position ? table.getState().columnPinning?.[position]?.indexOf(column.id) ?? -1 : 0;
    };
  },
  createTable: (table) => {
    table.setColumnPinning = (updater) => table.options.onColumnPinningChange?.(updater);
    table.resetColumnPinning = (defaultState) => table.setColumnPinning(
      defaultState ? getDefaultColumnPinningState() : table.initialState?.columnPinning ?? getDefaultColumnPinningState()
    );
    table.getIsSomeColumnsPinned = (position) => {
      const pinningState = table.getState().columnPinning;
      if (!position) {
        return Boolean(pinningState.left?.length || pinningState.right?.length);
      }
      return Boolean(pinningState[position]?.length);
    };
    table.getLeftLeafColumns = (0, import_utils.memo)(
      () => [table.getAllLeafColumns(), table.getState().columnPinning.left],
      (allColumns, left) => {
        return (left ?? []).map(
          (columnId) => allColumns.find((column) => column.id === columnId)
        ).filter(Boolean);
      },
      (0, import_utils.getMemoOptions)(table.options, "debugColumns", "getLeftLeafColumns")
    );
    table.getRightLeafColumns = (0, import_utils.memo)(
      () => [table.getAllLeafColumns(), table.getState().columnPinning.right],
      (allColumns, right) => {
        return (right ?? []).map(
          (columnId) => allColumns.find((column) => column.id === columnId)
        ).filter(Boolean);
      },
      (0, import_utils.getMemoOptions)(table.options, "debugColumns", "getRightLeafColumns")
    );
    table.getCenterLeafColumns = (0, import_utils.memo)(
      () => [
        table.getAllLeafColumns(),
        table.getState().columnPinning.left,
        table.getState().columnPinning.right
      ],
      (allColumns, left, right) => {
        const leftAndRight = [...left ?? [], ...right ?? []];
        return allColumns.filter((d) => !leftAndRight.includes(d.id));
      },
      (0, import_utils.getMemoOptions)(table.options, "debugColumns", "getCenterLeafColumns")
    );
    Object.assign((0, import__.getRowProto)(table), {
      getCenterVisibleCells: (0, import_utils.memo)(
        function() {
          return [
            this._getAllVisibleCells(),
            table.getState().columnPinning.left,
            table.getState().columnPinning.right
          ];
        },
        (allCells, left, right) => {
          const leftAndRight = [...left ?? [], ...right ?? []];
          return allCells.filter((d) => !leftAndRight.includes(d.column.id));
        },
        (0, import_utils.getMemoOptions)(table.options, "debugRows", "getCenterVisibleCells")
      ),
      getLeftVisibleCells: (0, import_utils.memo)(
        function() {
          return [
            this._getAllVisibleCells(),
            table.getState().columnPinning.left
          ];
        },
        (allCells, left) => {
          const cells = (left ?? []).map(
            (columnId) => allCells.find((cell) => cell.column.id === columnId)
          ).filter(Boolean).map((d) => ({ ...d, position: "left" }));
          return cells;
        },
        (0, import_utils.getMemoOptions)(table.options, "debugRows", "getLeftVisibleCells")
      ),
      getRightVisibleCells: (0, import_utils.memo)(
        function() {
          return [
            this._getAllVisibleCells(),
            table.getState().columnPinning.right
          ];
        },
        (allCells, right) => {
          const cells = (right ?? []).map(
            (columnId) => allCells.find((cell) => cell.column.id === columnId)
          ).filter(Boolean).map((d) => ({ ...d, position: "right" }));
          return cells;
        },
        (0, import_utils.getMemoOptions)(table.options, "debugRows", "getRightVisibleCells")
      )
    });
  }
};
