// @dynatrace/strato-components/tables/DataTable/tanstack-table-core-override/src/features/GlobalFiltering.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var GlobalFiltering_exports = {};
__export(GlobalFiltering_exports, {
  GlobalFiltering: () => GlobalFiltering
});
module.exports = __toCommonJS(GlobalFiltering_exports);
var import_filterFns = require("../filterFns");
var import_utils = require("../utils");
const GlobalFiltering = {
  getInitialState: (state) => {
    return {
      globalFilter: void 0,
      ...state
    };
  },
  getDefaultOptions: (table) => {
    return {
      onGlobalFilterChange: (0, import_utils.makeStateUpdater)("globalFilter", table),
      globalFilterFn: "auto",
      getColumnCanGlobalFilter: (column) => {
        const value = table.getCoreRowModel().flatRows[0]?._getAllCellsByColumnId()[column.id]?.getValue();
        return typeof value === "string" || typeof value === "number";
      }
    };
  },
  createColumn: (column, table) => {
    column.getCanGlobalFilter = () => {
      return (column.columnDef.enableGlobalFilter ?? true) && (table.options.enableGlobalFilter ?? true) && (table.options.enableFilters ?? true) && (table.options.getColumnCanGlobalFilter?.(column) ?? true) && !!column.accessorFn;
    };
  },
  createTable: (table) => {
    table.getGlobalAutoFilterFn = () => {
      return import_filterFns.filterFns.includesString;
    };
    table.getGlobalFilterFn = () => {
      const { globalFilterFn } = table.options;
      return (0, import_utils.isFunction)(globalFilterFn) ? globalFilterFn : globalFilterFn === "auto" ? table.getGlobalAutoFilterFn() : table.options.filterFns?.[globalFilterFn] ?? import_filterFns.filterFns[globalFilterFn];
    };
    table.setGlobalFilter = (updater) => {
      table.options.onGlobalFilterChange?.(updater);
    };
    table.resetGlobalFilter = (defaultState) => {
      table.setGlobalFilter(
        defaultState ? void 0 : table.initialState.globalFilter
      );
    };
  }
};
