// @dynatrace/strato-components/tables/DataTable/tanstack-table-core-override/src/features/RowExpanding.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var RowExpanding_exports = {};
__export(RowExpanding_exports, {
  RowExpanding: () => RowExpanding
});
module.exports = __toCommonJS(RowExpanding_exports);
var import__ = require("..");
var import_utils = require("../utils");
const RowExpanding = {
  getInitialState: (state) => {
    return {
      expanded: {},
      ...state
    };
  },
  getDefaultOptions: (table) => {
    return {
      onExpandedChange: (0, import_utils.makeStateUpdater)("expanded", table),
      paginateExpandedRows: true
    };
  },
  createTable: (table) => {
    let registered = false;
    let queued = false;
    table._autoResetExpanded = () => {
      if (!registered) {
        table._queue(() => {
          registered = true;
        });
        return;
      }
      if (table.options.autoResetAll ?? table.options.autoResetExpanded ?? !table.options.manualExpanding) {
        if (queued) return;
        queued = true;
        table._queue(() => {
          table.resetExpanded();
          queued = false;
        });
      }
    };
    table.setExpanded = (updater) => table.options.onExpandedChange?.(updater);
    table.toggleAllRowsExpanded = (expanded) => {
      if (expanded ?? !table.getIsAllRowsExpanded()) {
        table.setExpanded(true);
      } else {
        table.setExpanded({});
      }
    };
    table.resetExpanded = (defaultState) => {
      table.setExpanded(defaultState ? {} : table.initialState?.expanded ?? {});
    };
    table.getCanSomeRowsExpand = () => {
      return table.getPrePaginationRowModel().flatRows.some((row) => row.getCanExpand());
    };
    table.getToggleAllRowsExpandedHandler = () => {
      return (e) => {
        e.persist?.();
        table.toggleAllRowsExpanded();
      };
    };
    table.getIsSomeRowsExpanded = () => {
      const expanded = table.getState().expanded;
      return expanded === true || Object.values(expanded).some(Boolean);
    };
    table.getIsAllRowsExpanded = () => {
      const expanded = table.getState().expanded;
      if (typeof expanded === "boolean") {
        return expanded === true;
      }
      if (!Object.keys(expanded).length) {
        return false;
      }
      if (table.getRowModel().flatRows.some((row) => !row.getIsExpanded())) {
        return false;
      }
      return true;
    };
    table.getExpandedDepth = () => {
      let maxDepth = 0;
      const rowIds = table.getState().expanded === true ? Object.keys(table.getRowModel().rowsById) : Object.keys(table.getState().expanded);
      rowIds.forEach((id) => {
        const splitId = id.split(".");
        maxDepth = Math.max(maxDepth, splitId.length);
      });
      return maxDepth;
    };
    table.getPreExpandedRowModel = () => table.getSortedRowModel();
    table.getExpandedRowModel = () => {
      if (!table._getExpandedRowModel && table.options.getExpandedRowModel) {
        table._getExpandedRowModel = table.options.getExpandedRowModel(table);
      }
      if (table.options.manualExpanding || !table._getExpandedRowModel) {
        return table.getPreExpandedRowModel();
      }
      return table._getExpandedRowModel();
    };
    Object.assign((0, import__.getRowProto)(table), {
      toggleExpanded(expanded) {
        table.setExpanded((old) => {
          const exists = old === true ? true : !!old?.[this.id];
          let oldExpanded = {};
          if (old === true) {
            Object.keys(table.getRowModel().rowsById).forEach((rowId) => {
              oldExpanded[rowId] = true;
            });
          } else {
            oldExpanded = old;
          }
          expanded = expanded ?? !exists;
          if (!exists && expanded) {
            return {
              ...oldExpanded,
              [this.id]: true
            };
          }
          if (exists && !expanded) {
            const { [this.id]: _, ...rest } = oldExpanded;
            return rest;
          }
          return old;
        });
      },
      getIsExpanded() {
        const expanded = table.getState().expanded;
        return !!(table.options.getIsRowExpanded?.(this) ?? (expanded === true || expanded?.[this.id]));
      },
      getCanExpand() {
        return table.options.getRowCanExpand?.(this) ?? ((table.options.enableExpanding ?? true) && !!this.subRows?.length);
      },
      getIsAllParentsExpanded() {
        let isFullyExpanded = true;
        let currentRow = this;
        while (isFullyExpanded && currentRow.parentId) {
          currentRow = table.getRow(currentRow.parentId, true);
          isFullyExpanded = currentRow.getIsExpanded();
        }
        return isFullyExpanded;
      },
      getToggleExpandedHandler() {
        const canExpand = this.getCanExpand();
        return () => {
          if (!canExpand) return;
          this.toggleExpanded();
        };
      }
    });
  }
};
