// @dynatrace/strato-components/tables/DataTable/tanstack-table-core-override/src/features/RowPagination.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var RowPagination_exports = {};
__export(RowPagination_exports, {
  RowPagination: () => RowPagination
});
module.exports = __toCommonJS(RowPagination_exports);
var import_utils = require("../utils");
const defaultPageIndex = 0;
const defaultPageSize = 10;
const getDefaultPaginationState = () => ({
  pageIndex: defaultPageIndex,
  pageSize: defaultPageSize
});
const RowPagination = {
  getInitialState: (state) => {
    return {
      ...state,
      pagination: {
        ...getDefaultPaginationState(),
        ...state?.pagination
      }
    };
  },
  getDefaultOptions: (table) => {
    return {
      onPaginationChange: (0, import_utils.makeStateUpdater)("pagination", table)
    };
  },
  createTable: (table) => {
    let registered = false;
    let queued = false;
    table._autoResetPageIndex = () => {
      if (!registered) {
        table._queue(() => {
          registered = true;
        });
        return;
      }
      if (table.options.autoResetAll ?? table.options.autoResetPageIndex ?? !table.options.manualPagination) {
        if (queued) return;
        queued = true;
        table._queue(() => {
          table.resetPageIndex();
          queued = false;
        });
      }
    };
    table.setPagination = (updater) => {
      const safeUpdater = (old) => {
        let newState = (0, import_utils.functionalUpdate)(updater, old);
        return newState;
      };
      return table.options.onPaginationChange?.(safeUpdater);
    };
    table.resetPagination = (defaultState) => {
      table.setPagination(
        defaultState ? getDefaultPaginationState() : table.initialState.pagination ?? getDefaultPaginationState()
      );
    };
    table.setPageIndex = (updater) => {
      table.setPagination((old) => {
        let pageIndex = (0, import_utils.functionalUpdate)(updater, old.pageIndex);
        const maxPageIndex = typeof table.options.pageCount === "undefined" || table.options.pageCount === -1 ? Number.MAX_SAFE_INTEGER : table.options.pageCount - 1;
        pageIndex = Math.max(0, Math.min(pageIndex, maxPageIndex));
        return {
          ...old,
          pageIndex
        };
      });
    };
    table.resetPageIndex = (defaultState) => {
      table.setPageIndex(
        defaultState ? defaultPageIndex : table.initialState?.pagination?.pageIndex ?? defaultPageIndex
      );
    };
    table.resetPageSize = (defaultState) => {
      table.setPageSize(
        defaultState ? defaultPageSize : table.initialState?.pagination?.pageSize ?? defaultPageSize
      );
    };
    table.setPageSize = (updater) => {
      table.setPagination((old) => {
        const pageSize = Math.max(1, (0, import_utils.functionalUpdate)(updater, old.pageSize));
        const topRowIndex = old.pageSize * old.pageIndex;
        const pageIndex = Math.floor(topRowIndex / pageSize);
        return {
          ...old,
          pageIndex,
          pageSize
        };
      });
    };
    table.setPageCount = (updater) => table.setPagination((old) => {
      let newPageCount = (0, import_utils.functionalUpdate)(
        updater,
        table.options.pageCount ?? -1
      );
      if (typeof newPageCount === "number") {
        newPageCount = Math.max(-1, newPageCount);
      }
      return {
        ...old,
        pageCount: newPageCount
      };
    });
    table.getPageOptions = (0, import_utils.memo)(
      () => [table.getPageCount()],
      (pageCount) => {
        let pageOptions = [];
        if (pageCount && pageCount > 0) {
          pageOptions = [...new Array(pageCount)].fill(null).map((_, i) => i);
        }
        return pageOptions;
      },
      (0, import_utils.getMemoOptions)(table.options, "debugTable", "getPageOptions")
    );
    table.getCanPreviousPage = () => table.getState().pagination.pageIndex > 0;
    table.getCanNextPage = () => {
      const { pageIndex } = table.getState().pagination;
      const pageCount = table.getPageCount();
      if (pageCount === -1) {
        return true;
      }
      if (pageCount === 0) {
        return false;
      }
      return pageIndex < pageCount - 1;
    };
    table.previousPage = () => {
      return table.setPageIndex((old) => old - 1);
    };
    table.nextPage = () => {
      return table.setPageIndex((old) => {
        return old + 1;
      });
    };
    table.firstPage = () => {
      return table.setPageIndex(0);
    };
    table.lastPage = () => {
      return table.setPageIndex(table.getPageCount() - 1);
    };
    table.getPrePaginationRowModel = () => table.getExpandedRowModel();
    table.getPaginationRowModel = () => {
      if (!table._getPaginationRowModel && table.options.getPaginationRowModel) {
        table._getPaginationRowModel = table.options.getPaginationRowModel(table);
      }
      if (table.options.manualPagination || !table._getPaginationRowModel) {
        return table.getPrePaginationRowModel();
      }
      return table._getPaginationRowModel();
    };
    table.getPageCount = () => {
      return table.options.pageCount ?? Math.ceil(table.getRowCount() / table.getState().pagination.pageSize);
    };
    table.getRowCount = () => {
      return table.options.rowCount ?? table.getPrePaginationRowModel().rows.length;
    };
  }
};
