// @dynatrace/strato-components/tables/DataTable/tanstack-table-core-override/src/features/RowPinning.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var RowPinning_exports = {};
__export(RowPinning_exports, {
  RowPinning: () => RowPinning
});
module.exports = __toCommonJS(RowPinning_exports);
var import__ = require("..");
var import_utils = require("../utils");
const getDefaultRowPinningState = () => ({
  top: [],
  bottom: []
});
const RowPinning = {
  getInitialState: (state) => {
    return {
      rowPinning: getDefaultRowPinningState(),
      ...state
    };
  },
  getDefaultOptions: (table) => {
    return {
      onRowPinningChange: (0, import_utils.makeStateUpdater)("rowPinning", table)
    };
  },
  createTable: (table) => {
    table.setRowPinning = (updater) => table.options.onRowPinningChange?.(updater);
    table.resetRowPinning = (defaultState) => table.setRowPinning(
      defaultState ? getDefaultRowPinningState() : table.initialState?.rowPinning ?? getDefaultRowPinningState()
    );
    table.getIsSomeRowsPinned = (position) => {
      const pinningState = table.getState().rowPinning;
      if (!position) {
        return Boolean(pinningState.top?.length || pinningState.bottom?.length);
      }
      return Boolean(pinningState[position]?.length);
    };
    table._getPinnedRows = (visibleRows, pinnedRowIds, position) => {
      const rows = table.options.keepPinnedRows ?? true ? (
        //get all rows that are pinned even if they would not be otherwise visible
        //account for expanded parent rows, but not pagination or filtering
        (pinnedRowIds ?? []).map((rowId) => {
          const row = table.getRow(rowId, true);
          return row.getIsAllParentsExpanded() ? row : null;
        })
      ) : (
        //else get only visible rows that are pinned
        (pinnedRowIds ?? []).map(
          (rowId) => visibleRows.find((row) => row.id === rowId)
        )
      );
      return rows.filter(Boolean).map((d) => ({ ...d, position }));
    };
    table.getTopRows = (0, import_utils.memo)(
      () => [table.getRowModel().rows, table.getState().rowPinning.top],
      (allRows, topPinnedRowIds) => table._getPinnedRows(allRows, topPinnedRowIds, "top"),
      (0, import_utils.getMemoOptions)(table.options, "debugRows", "getTopRows")
    );
    table.getBottomRows = (0, import_utils.memo)(
      () => [table.getRowModel().rows, table.getState().rowPinning.bottom],
      (allRows, bottomPinnedRowIds) => table._getPinnedRows(allRows, bottomPinnedRowIds, "bottom"),
      (0, import_utils.getMemoOptions)(table.options, "debugRows", "getBottomRows")
    );
    table.getCenterRows = (0, import_utils.memo)(
      () => [
        table.getRowModel().rows,
        table.getState().rowPinning.top,
        table.getState().rowPinning.bottom
      ],
      (allRows, top, bottom) => {
        const topAndBottom = /* @__PURE__ */ new Set([...top ?? [], ...bottom ?? []]);
        return allRows.filter((d) => !topAndBottom.has(d.id));
      },
      (0, import_utils.getMemoOptions)(table.options, "debugRows", "getCenterRows")
    );
    Object.assign((0, import__.getRowProto)(table), {
      pin(position, includeLeafRows, includeParentRows) {
        const leafRowIds = includeLeafRows ? this.getLeafRows().map(({ id }) => id) : [];
        const parentRowIds = includeParentRows ? this.getParentRows().map(({ id }) => id) : [];
        const rowIds = /* @__PURE__ */ new Set([...parentRowIds, this.id, ...leafRowIds]);
        table.setRowPinning((old) => {
          if (position === "bottom") {
            return {
              top: (old?.top ?? []).filter((d) => !rowIds?.has(d)),
              bottom: [
                ...(old?.bottom ?? []).filter((d) => !rowIds?.has(d)),
                ...Array.from(rowIds)
              ]
            };
          }
          if (position === "top") {
            return {
              top: [
                ...(old?.top ?? []).filter((d) => !rowIds?.has(d)),
                ...Array.from(rowIds)
              ],
              bottom: (old?.bottom ?? []).filter((d) => !rowIds?.has(d))
            };
          }
          return {
            top: (old?.top ?? []).filter((d) => !rowIds?.has(d)),
            bottom: (old?.bottom ?? []).filter((d) => !rowIds?.has(d))
          };
        });
      },
      getCanPin() {
        const { enableRowPinning, enablePinning } = table.options;
        if (typeof enableRowPinning === "function") {
          return enableRowPinning(this);
        }
        return enableRowPinning ?? enablePinning ?? true;
      },
      getIsPinned() {
        const rowIds = [this.id];
        const { top, bottom } = table.getState().rowPinning;
        const isTop = rowIds.some((d) => top?.includes(d));
        const isBottom = rowIds.some((d) => bottom?.includes(d));
        return isTop ? "top" : isBottom ? "bottom" : false;
      },
      getPinnedIndex() {
        const position = this.getIsPinned();
        if (!position) return -1;
        const visiblePinnedRowIds = (position === "top" ? table.getTopRows() : table.getBottomRows())?.map(({ id }) => id);
        return visiblePinnedRowIds?.indexOf(this.id) ?? -1;
      }
    });
  }
};
