// @dynatrace/strato-components/tables/DataTable/tanstack-table-core-override/src/features/RowSorting.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var RowSorting_exports = {};
__export(RowSorting_exports, {
  RowSorting: () => RowSorting
});
module.exports = __toCommonJS(RowSorting_exports);
var import_sortingFns = require("../sortingFns");
var import_utils = require("../utils");
const RowSorting = {
  getInitialState: (state) => {
    return {
      sorting: [],
      ...state
    };
  },
  getDefaultColumnDef: () => {
    return {
      sortingFn: "auto",
      sortUndefined: 1
    };
  },
  getDefaultOptions: (table) => {
    return {
      onSortingChange: (0, import_utils.makeStateUpdater)("sorting", table),
      isMultiSortEvent: (e) => {
        return e.shiftKey;
      }
    };
  },
  createColumn: (column, table) => {
    column.getAutoSortingFn = () => {
      const firstRows = table.getFilteredRowModel().flatRows.slice(10);
      let isString = false;
      for (const row of firstRows) {
        const value = row?.getValue(column.id);
        if (Object.prototype.toString.call(value) === "[object Date]") {
          return import_sortingFns.sortingFns.datetime;
        }
        if (typeof value === "string") {
          isString = true;
          if (value.split(import_sortingFns.reSplitAlphaNumeric).length > 1) {
            return import_sortingFns.sortingFns.alphanumeric;
          }
        }
      }
      if (isString) {
        return import_sortingFns.sortingFns.text;
      }
      return import_sortingFns.sortingFns.basic;
    };
    column.getAutoSortDir = () => {
      const firstRow = table.getFilteredRowModel().flatRows[0];
      const value = firstRow?.getValue(column.id);
      if (typeof value === "string") {
        return "asc";
      }
      return "desc";
    };
    column.getSortingFn = () => {
      if (!column) {
        throw new Error();
      }
      return (0, import_utils.isFunction)(column.columnDef.sortingFn) ? column.columnDef.sortingFn : column.columnDef.sortingFn === "auto" ? column.getAutoSortingFn() : table.options.sortingFns?.[column.columnDef.sortingFn] ?? import_sortingFns.sortingFns[column.columnDef.sortingFn];
    };
    column.toggleSorting = (desc, multi) => {
      const nextSortingOrder = column.getNextSortingOrder();
      const hasManualValue = typeof desc !== "undefined" && desc !== null;
      table.setSorting((old) => {
        const existingSorting = old?.find((d) => d.id === column.id);
        const existingIndex = old?.findIndex((d) => d.id === column.id);
        let newSorting = [];
        let sortAction;
        let nextDesc = hasManualValue ? desc : nextSortingOrder === "desc";
        if (old?.length && column.getCanMultiSort() && multi) {
          if (existingSorting) {
            sortAction = "toggle";
          } else {
            sortAction = "add";
          }
        } else {
          if (old?.length && existingIndex !== old.length - 1) {
            sortAction = "replace";
          } else if (existingSorting) {
            sortAction = "toggle";
          } else {
            sortAction = "replace";
          }
        }
        if (sortAction === "toggle") {
          if (!hasManualValue) {
            if (!nextSortingOrder) {
              sortAction = "remove";
            }
          }
        }
        if (sortAction === "add") {
          newSorting = [
            ...old,
            {
              id: column.id,
              desc: nextDesc
            }
          ];
          newSorting.splice(
            0,
            newSorting.length - (table.options.maxMultiSortColCount ?? Number.MAX_SAFE_INTEGER)
          );
        } else if (sortAction === "toggle") {
          newSorting = old.map((d) => {
            if (d.id === column.id) {
              return {
                ...d,
                desc: nextDesc
              };
            }
            return d;
          });
        } else if (sortAction === "remove") {
          newSorting = old.filter((d) => d.id !== column.id);
        } else {
          newSorting = [
            {
              id: column.id,
              desc: nextDesc
            }
          ];
        }
        return newSorting;
      });
    };
    column.getFirstSortDir = () => {
      const sortDescFirst = column.columnDef.sortDescFirst ?? table.options.sortDescFirst ?? column.getAutoSortDir() === "desc";
      return sortDescFirst ? "desc" : "asc";
    };
    column.getNextSortingOrder = (multi) => {
      const firstSortDirection = column.getFirstSortDir();
      const isSorted = column.getIsSorted();
      if (!isSorted) {
        return firstSortDirection;
      }
      if (isSorted !== firstSortDirection && (table.options.enableSortingRemoval ?? true) && // If enableSortRemove, enable in general
      (multi ? table.options.enableMultiRemove ?? true : true)) {
        return false;
      }
      return isSorted === "desc" ? "asc" : "desc";
    };
    column.getCanSort = () => {
      return (column.columnDef.enableSorting ?? true) && (table.options.enableSorting ?? true) && !!column.accessorFn;
    };
    column.getCanMultiSort = () => {
      return column.columnDef.enableMultiSort ?? table.options.enableMultiSort ?? !!column.accessorFn;
    };
    column.getIsSorted = () => {
      const columnSort = table.getState().sorting?.find((d) => d.id === column.id);
      return !columnSort ? false : columnSort.desc ? "desc" : "asc";
    };
    column.getSortIndex = () => table.getState().sorting?.findIndex((d) => d.id === column.id) ?? -1;
    column.clearSorting = () => {
      table.setSorting(
        (old) => old?.length ? old.filter((d) => d.id !== column.id) : []
      );
    };
    column.getToggleSortingHandler = () => {
      const canSort = column.getCanSort();
      return (e) => {
        if (!canSort) return;
        e.persist?.();
        column.toggleSorting?.(
          void 0,
          column.getCanMultiSort() ? table.options.isMultiSortEvent?.(e) : false
        );
      };
    };
  },
  createTable: (table) => {
    table.setSorting = (updater) => table.options.onSortingChange?.(updater);
    table.resetSorting = (defaultState) => {
      table.setSorting(defaultState ? [] : table.initialState?.sorting ?? []);
    };
    table.getPreSortedRowModel = () => table.getGroupedRowModel();
    table.getSortedRowModel = () => {
      if (!table._getSortedRowModel && table.options.getSortedRowModel) {
        table._getSortedRowModel = table.options.getSortedRowModel(table);
      }
      if (table.options.manualSorting || !table._getSortedRowModel) {
        return table.getPreSortedRowModel();
      }
      return table._getSortedRowModel();
    };
  }
};
