// @dynatrace/strato-components/esm/charts/core/components/legend/range/components/RangeLegendMarkers/RangeLegendMarkers.js
import { jsx, jsxs } from "react/jsx-runtime";
import { isNumber } from "lodash-es";
import { Fragment, useMemo } from "react";
import { legendMarkersSvg } from "./RangeLegendMarkers.sty.js";
import { ORIENTATION } from "../../../../../constants.js";
import { LABEL_ALIGNMENT } from "../../constants.js";
import { useOrientation } from "../../context/RangeLegendOrientationContext.js";
import { useScale } from "../../context/RangeLegendScaleContext.js";
import { useLegendFormattedLabels } from "../../hooks/use-legend-formatted-labels.js";
import { getLabelAlignment } from "../../utils/get-label-alignment.js";
import { getLabelEndPosition } from "../../utils/get-label-end-position.js";
import { getMarkers } from "../../utils/get-markers.js";
import { RangeLegendBoundaryLabels } from "../RangeLegendBoundaryLabels.js";
import { RangeLegendItemLabel } from "../RangeLegendItemLabel/RangeLegendItemLabel.js";
import { RangeLegendItemMarker } from "../RangeLegendItemMarker/RangeLegendItemMarker.js";
import { RangeLegendSegmentsOverlay } from "../RangeLegendSegmentsOverlay.js";
const RangeLegendMarkers = (props) => {
  const { items, maxValue, minValue } = props;
  const hasMinValue = isNumber(minValue);
  const hasMaxValue = isNumber(maxValue);
  const { orientation } = useOrientation();
  const { scale } = useScale();
  const [rangeStartPosition, rangeEndPosition] = scale.range();
  const isVertical = orientation === ORIENTATION.Vertical;
  const { itemMarkers, singlePointRangeMarkers } = useMemo(
    () => getMarkers(items, isVertical),
    [isVertical, items]
  );
  const internalRangeItems = useLegendFormattedLabels(items);
  const addedMarkers = /* @__PURE__ */ new Set([]);
  const markerLabel = isVertical ? `Labels from ${Array.from(itemMarkers)[itemMarkers.size - 1]} to ${Array.from(itemMarkers)[0]}` : `Labels from ${Array.from(itemMarkers)[0]} to ${Array.from(itemMarkers)[itemMarkers.size - 1]}`;
  return /* @__PURE__ */ jsxs(
    "svg",
    {
      className: legendMarkersSvg({
        isHorizontal: orientation === ORIENTATION.Horizontal
      }),
      role: "group",
      "aria-label": "Chart legend",
      children: [
        /* @__PURE__ */ jsxs("g", { role: "list", "aria-label": markerLabel, children: [
          internalRangeItems.map((item) => {
            const hasFromLabel = itemMarkers.has(item.from) && !addedMarkers.has(item.from);
            const hasToLabel = itemMarkers.has(item.to) && !addedMarkers.has(item.to);
            if (!hasFromLabel && !hasToLabel) {
              return null;
            }
            addedMarkers.add(item.from);
            addedMarkers.add(item.to);
            const startPosition = scale(item.from);
            const endPosition = scale(item.to);
            const isSinglePointRange = startPosition === endPosition;
            const hasSameStartPosition = isVertical ? startPosition === rangeEndPosition : startPosition === rangeStartPosition;
            const isFirstItem = hasSameStartPosition && !hasMinValue;
            const hasSameEndPosition = isVertical ? endPosition === rangeStartPosition : endPosition === rangeEndPosition;
            const isLastItem = hasSameEndPosition && !hasMaxValue;
            const labelAlignment = getLabelAlignment(
              hasMaxValue,
              isLastItem,
              isVertical
            );
            const labelEndPosition = getLabelEndPosition(
              endPosition,
              isSinglePointRange,
              isLastItem
            );
            return /* @__PURE__ */ jsxs(Fragment, { children: [
              hasFromLabel && /* @__PURE__ */ jsx(
                RangeLegendItemLabel,
                {
                  labelId: item.from,
                  startPosition,
                  endPosition: startPosition,
                  text: item.labels.from,
                  alignment: hasMinValue || !isFirstItem ? LABEL_ALIGNMENT.Center : LABEL_ALIGNMENT.End
                }
              ),
              hasToLabel && !isSinglePointRange && /* @__PURE__ */ jsx(
                RangeLegendItemLabel,
                {
                  labelId: item.to,
                  startPosition: labelEndPosition,
                  endPosition: labelEndPosition,
                  text: item.labels.to,
                  alignment: labelAlignment
                }
              )
            ] }, `label-'${item.from}`);
          }),
          Array.from(itemMarkers).map((marker) => {
            const position = scale(marker);
            return /* @__PURE__ */ jsx(
              RangeLegendItemMarker,
              {
                position,
                markerValue: marker,
                isReversed: isVertical,
                isSinglePointRange: singlePointRangeMarkers.has(marker)
              },
              `marker-'${marker}`
            );
          }),
          /* @__PURE__ */ jsx(
            RangeLegendBoundaryLabels,
            {
              items: internalRangeItems,
              maxValue,
              minValue
            }
          )
        ] }),
        /* @__PURE__ */ jsx(RangeLegendSegmentsOverlay, { itemMarkers: Array.from(itemMarkers) })
      ]
    }
  );
};
export {
  RangeLegendMarkers
};
