// @dynatrace/strato-geo/esm/map/components/toolbar/buttons/MapDownloadDataButton.js
import { jsx } from "react/jsx-runtime";
import { useRef } from "react";
import { useIntl } from "react-intl";
import { _ChartToolbarRenderer as ChartToolbarRenderer } from "@dynatrace/strato-components/charts";
import { DownloadIcon } from "@dynatrace/strato-icons";
import { useMapRawData } from "../../../hooks/use-map-raw-data.js";
import { buildAndDownloadCsv } from "../../../utils/build-and-download-csv.js";
import { TOOLBAR_MESSAGES } from "../constants.js";
import { useA11yToolbarTabIndex } from "../hooks/use-a11y-toolbar-tab-index.js";
const MapDownloadDataButton = () => {
  const downloadDataButtonRef = useRef(null);
  const intl = useIntl();
  const tabIndex = useA11yToolbarTabIndex("download-data");
  const data = useMapRawData();
  const handleDownloadDataKeyboard = (event) => {
    if (event.key === "Enter") {
      buildAndDownloadCsv(data);
    }
  };
  const handleDownloadDataToolbar = () => {
    buildAndDownloadCsv(data);
  };
  return /* @__PURE__ */ jsx(
    ChartToolbarRenderer.Control,
    {
      ref: downloadDataButtonRef,
      prefixIcon: /* @__PURE__ */ jsx(DownloadIcon, {}),
      tooltipText: intl.formatMessage(TOOLBAR_MESSAGES.downloadData),
      testId: "map-toolbar-download",
      onKeyDown: handleDownloadDataKeyboard,
      onClick: handleDownloadDataToolbar,
      tabIndex,
      placement: "toolbar"
    }
  );
};
export {
  MapDownloadDataButton
};
