// @dynatrace/strato-geo/esm/map/components/toolbar/buttons/MapResetButton.js
import { jsx } from "react/jsx-runtime";
import { useMap } from "@vis.gl/react-maplibre";
import { useRef } from "react";
import { useIntl } from "react-intl";
import { _ChartToolbarRenderer as ChartToolbarRenderer } from "@dynatrace/strato-components/charts";
import { ResetIcon } from "@dynatrace/strato-icons";
import { useInitialViewContext } from "../../../hooks/use-initial-view-context.js";
import { useMapZoomState, useSetStateZoom } from "../../../store/store.js";
import { TOOLBAR_MESSAGES } from "../constants.js";
import { useA11yToolbarTabIndex } from "../hooks/use-a11y-toolbar-tab-index.js";
const MapResetButton = () => {
  const { current: map } = useMap();
  const { longitude = 0, latitude = 0, zoom = 0 } = useInitialViewContext();
  const intl = useIntl();
  const { resetEnabled } = useMapZoomState();
  const setZoomState = useSetStateZoom();
  const resetButtonRef = useRef(null);
  const tabIndex = useA11yToolbarTabIndex("reset");
  const handleReset = () => {
    map?.flyTo({ center: [longitude, latitude], zoom });
    setZoomState({ zoomToFitEnabled: true, resetEnabled: false });
  };
  const handleKeyDown = (event) => {
    if (event.code === "Enter") {
      handleReset();
    }
  };
  return /* @__PURE__ */ jsx(
    ChartToolbarRenderer.Control,
    {
      ref: resetButtonRef,
      prefixIcon: /* @__PURE__ */ jsx(ResetIcon, {}),
      tooltipText: intl.formatMessage(TOOLBAR_MESSAGES.zoomActionReset),
      testId: "map-toolbar-reset",
      onKeyDown: handleKeyDown,
      onClick: handleReset,
      tabIndex,
      disabled: !resetEnabled
    }
  );
};
export {
  MapResetButton
};
