// @dynatrace/strato-geo/esm/map/components/toolbar/buttons/MapZoomInOutButtons.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { useMap } from "@vis.gl/react-maplibre";
import { useRef } from "react";
import { useIntl } from "react-intl";
import { _ChartToolbarRenderer as ChartToolbarRenderer } from "@dynatrace/strato-components/charts";
import { ZoomInIcon, ZoomOutIcon } from "@dynatrace/strato-icons";
import { useMapZoomState } from "../../../store/selectors.js";
import { TOOLBAR_MESSAGES } from "../constants.js";
import { useA11yToolbarTabIndex } from "../hooks/use-a11y-toolbar-tab-index.js";
const MapZoomInOutButtons = () => {
  const { current: map } = useMap();
  const { zoomInEnabled, zoomOutEnabled } = useMapZoomState();
  const zoomInButtonRef = useRef(null);
  const zoomButtonOutRef = useRef(null);
  const intl = useIntl();
  const tabIndexIn = useA11yToolbarTabIndex("zoom-in");
  const tabIndexOut = useA11yToolbarTabIndex("zoom-out");
  const handleZoomKeyboardMode = (event, zoomBehavior) => {
    if (event.code === "Enter") {
      handleZoomStepToolbarMode(zoomBehavior);
    }
  };
  const handleZoomStepToolbarMode = (zoomBehavior) => {
    if (zoomBehavior === "zoom-in") {
      map?.zoomIn();
    }
    if (zoomBehavior === "zoom-out") {
      map?.zoomOut();
    }
  };
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(
      ChartToolbarRenderer.Control,
      {
        ref: zoomInButtonRef,
        prefixIcon: /* @__PURE__ */ jsx(ZoomInIcon, {}),
        tooltipText: intl.formatMessage(TOOLBAR_MESSAGES.zoomActionZoomIn),
        testId: "map-toolbar-zoom-in",
        onKeyDown: (event) => handleZoomKeyboardMode(event, "zoom-in"),
        onClick: () => handleZoomStepToolbarMode("zoom-in"),
        tabIndex: tabIndexIn,
        disabled: !zoomInEnabled
      }
    ),
    /* @__PURE__ */ jsx(
      ChartToolbarRenderer.Control,
      {
        ref: zoomButtonOutRef,
        prefixIcon: /* @__PURE__ */ jsx(ZoomOutIcon, {}),
        tooltipText: intl.formatMessage(TOOLBAR_MESSAGES.zoomActionZoomOut),
        testId: "map-toolbar-zoom-out",
        onKeyDown: (event) => handleZoomKeyboardMode(event, "zoom-out"),
        onClick: () => handleZoomStepToolbarMode("zoom-out"),
        tabIndex: tabIndexOut,
        disabled: !zoomOutEnabled
      }
    )
  ] });
};
export {
  MapZoomInOutButtons
};
