// @dynatrace/strato-geo/esm/map/components/toolbar/buttons/MapZoomToFitButton.js
import { jsx } from "react/jsx-runtime";
import { useMap } from "@vis.gl/react-maplibre";
import { useRef } from "react";
import { useIntl } from "react-intl";
import { _ChartToolbarRenderer as ChartToolbarRenderer } from "@dynatrace/strato-components/charts";
import { ZoomToFitIcon } from "@dynatrace/strato-icons";
import { FIT_BOUNDS_OPTIONS } from "../../../constants.js";
import { useDataBoundingBox } from "../../../hooks/use-data-bounding-box.js";
import { useMapZoomState, useSetStateZoom } from "../../../store/store.js";
import { TOOLBAR_MESSAGES } from "../constants.js";
import { useA11yToolbarTabIndex } from "../hooks/use-a11y-toolbar-tab-index.js";
const MapZoomToFitButton = () => {
  const resetButtonRef = useRef(null);
  const { current: map } = useMap();
  const tabIndex = useA11yToolbarTabIndex("zoom-to-fit");
  const intl = useIntl();
  const boundingBox = useDataBoundingBox();
  const { zoomToFitEnabled } = useMapZoomState();
  const setZoomState = useSetStateZoom();
  const handleZoomToFit = () => {
    map?.fitBounds(boundingBox, FIT_BOUNDS_OPTIONS);
    setZoomState({ zoomToFitEnabled: false, zoomToFitInProgress: true });
  };
  const handleKeyDown = (event) => {
    if (event.code === "Enter") {
      handleZoomToFit();
    }
  };
  return /* @__PURE__ */ jsx(
    ChartToolbarRenderer.Control,
    {
      ref: resetButtonRef,
      prefixIcon: /* @__PURE__ */ jsx(ZoomToFitIcon, {}),
      tooltipText: intl.formatMessage(TOOLBAR_MESSAGES.zoomActionZoomToFit),
      testId: "map-toolbar-zoom-to-fit",
      onKeyDown: handleKeyDown,
      onClick: handleZoomToFit,
      tabIndex,
      disabled: !zoomToFitEnabled
    }
  );
};
MapZoomToFitButton["displayName"] = "MapZoomToFit";
export {
  MapZoomToFitButton
};
