// @dynatrace/strato-components/forms/spin-buttons/hooks/useSpinButtonInteraction.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useSpinButtonInteraction_exports = {};
__export(useSpinButtonInteraction_exports, {
  useSpinButtonInteraction: () => useSpinButtonInteraction
});
module.exports = __toCommonJS(useSpinButtonInteraction_exports);
function useSpinButtonInteraction(props) {
  const { disabled, selectAll, deselectAll, rootElement } = props;
  return {
    onClick: (event) => {
      if (disabled) {
        event.preventDefault();
      } else {
        selectAll();
      }
    },
    onFocus: (event) => {
      if (event.target.hasAttribute("contenteditable") === false || event.target.getAttribute("contenteditable") === "false") {
        return;
      }
      const selection = window.getSelection();
      const selectionSpansMultipleNodes = selection && selection.anchorNode !== selection.focusNode && selection.toString().length > 0;
      const isRestoringSelection = event.target.closest(".strato-inputgroup")?.hasAttribute("data-restoring-selection") ?? false;
      if (selectionSpansMultipleNodes || isRestoringSelection) {
        return;
      }
      if (disabled) {
        event.preventDefault();
      } else {
        selectAll();
        rootElement?.scrollIntoView({
          block: "nearest",
          inline: "nearest"
        });
      }
    },
    onBlur: (event) => {
      if (event.target.hasAttribute("contenteditable") === false || event.target.getAttribute("contenteditable") === "false") {
        return;
      }
      deselectAll();
    }
  };
}
