// @dynatrace/strato-components/esm/charts/xy-chart/processing-strategies/utils/stacking/stack-datapoints.js
import { isCategoricalAxis } from "../../../types/guards.js";
import {
  BAR_VARIANT_STACK
} from "../../../types/xy-chart.js";
const orientationToBaselineAxis = {
  vertical: "x",
  horizontal: "y"
};
const getCombinedAxesOrientationKey = (xAxisId, yAxisId, chartOrientation) => `${xAxisId}-${yAxisId}-${chartOrientation}`;
const getStackingData = (chartOrientation, datapoint, xAxis, yAxis) => {
  const [start, end] = chartOrientation === "vertical" ? [datapoint.x0, datapoint.x1] : [datapoint.y0, datapoint.y1];
  const stackKey = (end ? `${start}-${end}` : `${start}`) + `-${getCombinedAxesOrientationKey(xAxis.id, yAxis.id, chartOrientation)}`;
  const value = chartOrientation === "vertical" ? datapoint.y1 : datapoint.x1;
  return { stackKey, value };
};
const getBaselineCoordinate = (orientation) => {
  if (!orientation || orientation === "both") {
    return;
  }
  return orientationToBaselineAxis[orientation];
};
const isStackable = (variant, chartOrientation, xAxis, yAxis) => {
  const baselineCoordinate = getBaselineCoordinate(chartOrientation);
  const isStackingAxisCategorical = baselineCoordinate && baselineCoordinate === "x" ? isCategoricalAxis(yAxis) : isCategoricalAxis(xAxis);
  return chartOrientation !== "both" && variant === BAR_VARIANT_STACK && !isStackingAxisCategorical;
};
const getStackedValue = (value, prevValue) => {
  return {
    minBoundary: value < 0 ? prevValue.minBoundary + value : prevValue.minBoundary,
    maxBoundary: value > 0 ? prevValue.maxBoundary + value : prevValue.maxBoundary
  };
};
export {
  getBaselineCoordinate,
  getCombinedAxesOrientationKey,
  getStackedValue,
  getStackingData,
  isStackable
};
