// @dynatrace/strato-components/charts/core/components/tooltip/ChartTooltip/components/TimestampHeader/TimestampHeader.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TimestampHeader_exports = {};
__export(TimestampHeader_exports, {
  TimestampHeader: () => TimestampHeader
});
module.exports = __toCommonJS(TimestampHeader_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_spacings = __toESM(require("@dynatrace/strato-design-tokens/spacings"));
var import_units = require("@dynatrace-sdk/units");
var import_TooltipItem = require("../../../section-components/TooltipItem/TooltipItem.js");
var import_TextHeader = require("../TextHeader.js");
const TimestampHeader = ({ date }) => {
  const formattedDate = (0, import_units.formatDate)(date, {
    day: "2-digit",
    month: "short",
    year: date.getFullYear() === (/* @__PURE__ */ new Date()).getFullYear() ? void 0 : "numeric",
    hour: "2-digit",
    minute: "2-digit",
    second: date.getSeconds() === 0 ? void 0 : "numeric"
  });
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TooltipItem.TooltipItem, { style: { padding: `0 ${import_spacings.default.Size2}` }, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TextHeader.TextHeader, { text: formattedDate }) });
};
TimestampHeader["displayName"] = "ChartTooltip.TimestampHeader";
