// @dynatrace/strato-components/esm/layouts/app-header/ActionItems.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { createElement } from "react";
import {
  forwardRef,
  useMemo,
  useRef,
  Children,
  Fragment as Fragment2,
  cloneElement
} from "react";
import { useIntl } from "react-intl";
import { isElement } from "react-is";
import { DotMenuIcon } from "@dynatrace/strato-icons";
import {
  ActionButton,
  isAppHeaderActionButton
} from "./ActionButton.js";
import { ActionItemGroup } from "./ActionItemGroup.js";
import { Button } from "../../buttons/button/Button.js";
import { useContainerSize } from "../../core/hooks/useContainerSize.js";
import { isSlot } from "../../core/utils/is-slot.js";
import { Menu } from "../../navigation/menu/Menu.js";
import { Tooltip } from "../../overlays/tooltip/Tooltip.js";
import { Divider } from "../divider/Divider.js";
import { Flex } from "../flex/Flex.js";
function getSlots(children) {
  const actions = [];
  const actionGroups = {};
  const custom = [];
  let groupCounter = 0;
  for (const child of Children.toArray(children)) {
    if (!isElement(child)) {
      continue;
    }
    const childWithKey = cloneElement(child, {
      key: child.key ?? crypto.randomUUID()
    });
    if (isSlot(child, ActionItemGroup)) {
      groupCounter++;
      actionGroups[groupCounter] = [];
      const actionGroup = actionGroups[groupCounter];
      const validChildren = Children.toArray(child.props.children).filter(
        (validChild) => isSlot(validChild, Tooltip) && isSlot(validChild.props.children, ActionButton) || isSlot(validChild, ActionButton)
      );
      validChildren.forEach((validChild) => actionGroup.push(validChild));
    } else if (isAppHeaderActionButton(child)) {
      actions.push(childWithKey);
    } else if (isSlot(child, Tooltip) && isSlot(child.props.children, ActionButton)) {
      actions.push(child);
    } else {
      custom.push(childWithKey);
    }
  }
  return { actions, actionGroups, custom };
}
function useCollapse() {
  const containerRef = useRef(null);
  const contentRef = useRef(null);
  const containerSize = useContainerSize(containerRef);
  const contentSize = useContainerSize(contentRef);
  return {
    isOverflowing: contentSize.width > containerSize.width,
    containerRef,
    contentRef
  };
}
const ActionItems = forwardRef(({ children, className, style, "data-testid": dataTestId, ...rest }, ref) => {
  const slots = useMemo(() => getSlots(children), [children]);
  const collapse = useCollapse();
  const intl = useIntl();
  const hasActions = slots.actions.length > 0;
  const hasGroups = Object.entries(slots.actionGroups).length > 0;
  const isCollapsedIntoMenu = collapse.isOverflowing && (hasActions || hasGroups);
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(
      Flex,
      {
        ref,
        alignItems: "center",
        justifyContent: "flex-end",
        flexGrow: 1,
        flexShrink: 0,
        flexBasis: hasActions || hasGroups ? 0 : "auto",
        minWidth: 0,
        style,
        className,
        "data-testid": dataTestId,
        children: /* @__PURE__ */ jsx(
          Flex,
          {
            ...rest,
            ref: collapse.containerRef,
            justifyContent: "flex-end",
            alignItems: "center",
            flexGrow: 1,
            minWidth: 0,
            hidden: collapse.isOverflowing,
            style: {
              visibility: collapse.isOverflowing ? "hidden" : void 0,
              flexBasis: collapse.isOverflowing ? 0 : void 0
            },
            children: /* @__PURE__ */ jsxs(
              Flex,
              {
                gap: 4,
                style: { minWidth: "max-content" },
                ref: collapse.contentRef,
                alignItems: "center",
                children: [
                  collapse.isOverflowing ? null : slots.custom,
                  slots.actions.map((action) => {
                    if (isSlot(action, ActionButton)) {
                      const { prefixIcon, children: children2, showLabel, ...rest2 } = action.props;
                      return /* @__PURE__ */ createElement(Button, { ...rest2, key: action.key, variant: "emphasized" }, prefixIcon ? /* @__PURE__ */ jsx(Button.Prefix, { children: prefixIcon }) : null, children2);
                    } else if (isSlot(action, Tooltip) && isSlot(action.props.children, ActionButton)) {
                      return action;
                    }
                    return null;
                  }),
                  hasActions && hasGroups ? /* @__PURE__ */ jsx(
                    Divider,
                    {
                      orientation: "vertical",
                      flexItem: false,
                      style: { height: 20 }
                    }
                  ) : null,
                  Object.entries(slots.actionGroups).map(
                    ([key, actions], actionGroupIndex) => /* @__PURE__ */ jsxs(Fragment2, { children: [
                      actionGroupIndex !== 0 && /* @__PURE__ */ jsx(
                        Divider,
                        {
                          orientation: "vertical",
                          flexItem: false,
                          style: { height: 20 }
                        }
                      ),
                      actions.map((action) => {
                        if (isSlot(action, ActionButton)) {
                          const { prefixIcon, children: children2, showLabel, ...rest2 } = action.props;
                          return /* @__PURE__ */ createElement(Button, { ...rest2, key: action.key, variant: "emphasized" }, prefixIcon ? /* @__PURE__ */ jsx(Button.Prefix, { children: prefixIcon }) : null, children2);
                        } else if (isSlot(action, Tooltip) && isSlot(action.props.children, ActionButton)) {
                          return action;
                        }
                        return null;
                      })
                    ] }, key)
                  )
                ]
              }
            )
          }
        )
      }
    ),
    collapse.isOverflowing ? /* @__PURE__ */ jsx(Flex, { gap: 4, alignItems: "center", children: slots.custom }) : null,
    isCollapsedIntoMenu ? /* @__PURE__ */ jsxs(Menu, { children: [
      /* @__PURE__ */ jsx(Menu.Trigger, { children: /* @__PURE__ */ jsx(
        Button,
        {
          variant: "default",
          "aria-label": intl.formatMessage({
            id: "L/D6X7Ee5PYWzl3i",
            defaultMessage: "Show more actions"
          }),
          children: /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(DotMenuIcon, {}) })
        }
      ) }),
      /* @__PURE__ */ jsxs(Menu.Content, { children: [
        /* @__PURE__ */ jsx(Menu.Group, { children: slots.actions.map((action) => {
          if (isSlot(action, ActionButton)) {
            const { prefixIcon, children: children2, showLabel, onClick, ...rest2 } = action.props;
            return /* @__PURE__ */ createElement(
              Menu.Item,
              {
                ...rest2,
                key: action.key,
                onSelect: onClick ? onClick : () => void 0
              },
              prefixIcon ? /* @__PURE__ */ jsx(Menu.Prefix, { children: prefixIcon }) : null,
              children2
            );
          } else if (isSlot(action, Tooltip) && isSlot(action.props.children, ActionButton)) {
            return /* @__PURE__ */ jsxs(
              Menu.Item,
              {
                onSelect: action.props.children.props.onClick ? action.props.children.props.onClick : () => void 0,
                ...action.props.children,
                children: [
                  action.props.children.props.prefixIcon ? /* @__PURE__ */ jsx(Menu.Prefix, { children: action.props.children.props.prefixIcon }) : null,
                  action.props.children.props.children,
                  /* @__PURE__ */ jsx(Menu.ItemTooltip, { children: action.props.text })
                ]
              }
            );
          }
          return null;
        }) }),
        Object.entries(slots.actionGroups).map(([groupId, actions]) => /* @__PURE__ */ jsx(Menu.Group, { children: actions.map((action) => {
          if (isSlot(action, ActionButton)) {
            const {
              prefixIcon,
              children: children2,
              showLabel,
              onClick,
              ...rest2
            } = action.props;
            return /* @__PURE__ */ createElement(
              Menu.Item,
              {
                ...rest2,
                key: action.key,
                onSelect: onClick ? onClick : () => void 0
              },
              prefixIcon ? /* @__PURE__ */ jsx(Menu.Prefix, { children: prefixIcon }) : null,
              children2
            );
          } else if (isSlot(action, Tooltip) && isSlot(action.props.children, ActionButton)) {
            return /* @__PURE__ */ jsxs(
              Menu.Item,
              {
                onSelect: action.props.children.props.onClick ? action.props.children.props.onClick : () => void 0,
                ...action.props.children,
                children: [
                  action.props.children.props.prefixIcon ? /* @__PURE__ */ jsx(Menu.Prefix, { children: action.props.children.props.prefixIcon }) : null,
                  action.props.children.props.children,
                  /* @__PURE__ */ jsx(Menu.ItemTooltip, { children: action.props.text })
                ]
              }
            );
          }
          return null;
        }) }, groupId))
      ] })
    ] }) : null
  ] });
});
ActionItems.displayName = "AppHeader.ActionItems";
export {
  ActionItems
};
