// @dynatrace/strato-components/esm/layouts/app-header/AppHeader.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import { Children, forwardRef, useMemo } from "react";
import { ActionButton } from "./ActionButton.js";
import { ActionItemGroup } from "./ActionItemGroup.js";
import { ActionItems } from "./ActionItems.js";
import { AppIcon } from "./app-icon/AppIcon.js";
import * as styles from "./AppHeader.sty.js";
import { Logo } from "./Logo.js";
import { Menus } from "./Menus.js";
import { Navigation, NavigationItem } from "./Navigation.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { OverlayStackingProvider } from "../../core/providers/OverlayStackingProvider.js";
import { isDev } from "../../core/utils/is-node-environment.js";
import { isSlot } from "../../core/utils/is-slot.js";
import { Divider } from "../divider/Divider.js";
import { HelpMenu } from "../help-menu/HelpMenu.js";
const COMPONENT_NAME = "AppHeader";
function getItemSlots(children) {
  const slots = {
    navigation: /* @__PURE__ */ jsx(AppHeader.Navigation, {})
  };
  for (const child of Children.toArray(children)) {
    if (isSlot(child, ActionItems)) {
      slots.actions = child;
      continue;
    }
    if (isSlot(child, HelpMenu)) {
      slots.menus = /* @__PURE__ */ jsxs(Fragment, { children: [
        slots.menus,
        child
      ] });
      continue;
    }
    if (isSlot(child, Menus)) {
      slots.menus = /* @__PURE__ */ jsxs(Fragment, { children: [
        child,
        slots.menus
      ] });
      continue;
    }
    if (isSlot(child, Navigation)) {
      slots.navigation = child;
      continue;
    }
    if (isDev()) {
      console.warn("The used child is not supported by the AppHeader.");
    }
  }
  return slots;
}
const AppHeaderRoot = forwardRef(
  (props, forwardedRef) => {
    const {
      children,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...remainingProps
    } = props;
    const slots = useMemo(() => getItemSlots(children), [children]);
    const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
    return /* @__PURE__ */ jsx(
      "header",
      {
        ...restProps,
        ...behavioralTrackingProps,
        ref: forwardedRef,
        className: clsx(styles.header, consumerClassName),
        style: consumerStyle,
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        "data-dt-component": COMPONENT_NAME,
        children: /* @__PURE__ */ jsxs(OverlayStackingProvider, { children: [
          slots.navigation,
          slots.actions,
          slots.actions && slots.menus ? /* @__PURE__ */ jsx(
            Divider,
            {
              orientation: "vertical",
              flexItem: false,
              style: { height: 20 }
            }
          ) : null,
          slots.menus
        ] })
      }
    );
  }
);
AppHeaderRoot.displayName = "AppHeader";
const AppHeader = Object.assign(AppHeaderRoot, {
  // DEPRECATION: [APPDEV-14264] BREAKING-CHANGE: AppHeader: Remove old NavItems API.
  /**
   * @deprecated use {@link AppHeader.Logo}
   */
  AppNavLink: Logo,
  // DEPRECATION: [APPDEV-14264] BREAKING-CHANGE: AppHeader: Remove old NavItems API.
  /**
   * @deprecated use {@link AppHeader.Navigation}
   */
  NavItems: Navigation,
  // DEPRECATION: [APPDEV-14264] BREAKING-CHANGE: AppHeader: Remove old NavItems API.
  /**
   * @deprecated use {@link AppHeader.NavigationItem}
   */
  NavItem: NavigationItem,
  AppIcon,
  ActionItems,
  ActionItemGroup,
  ActionButton,
  Logo,
  Menus,
  Navigation,
  NavigationItem
});
export {
  AppHeader
};
