// @dynatrace/strato-components/esm/layouts/app-header/Navigation.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { createElement } from "react";
import clsx from "clsx";
import {
  forwardRef,
  useMemo,
  useRef,
  Children
} from "react";
import { useIntl } from "react-intl";
import { isElement } from "react-is";
import { MenuIcon } from "@dynatrace/strato-icons";
import { Logo } from "./Logo.js";
import * as styles from "./Navigation.sty.js";
import { Button } from "../../buttons/button/Button.js";
import { useContainerSize } from "../../core/hooks/useContainerSize.js";
import { isSlot } from "../../core/utils/is-slot.js";
import { Menu } from "../../navigation/menu/Menu.js";
import { Flex } from "../flex/Flex.js";
function getSlots(children) {
  const linksProps = [];
  let logo;
  for (const child of Children.toArray(children)) {
    if (!isElement(child)) {
      continue;
    }
    if (isSlot(child, NavigationItem)) {
      const key = child.key ?? crypto.randomUUID();
      const propsWithKey = Object.assign({}, child.props, { key });
      linksProps.push(propsWithKey);
    } else if (isSlot(child, Logo)) {
      logo = child;
    }
  }
  return { linksProps, logo };
}
function useCollapsingNav() {
  const containerRef = useRef(null);
  const contentRef = useRef(null);
  const containerSize = useContainerSize(containerRef);
  const contentSize = useContainerSize(contentRef);
  return {
    isOverflowing: contentSize.width > containerSize.width,
    containerRef,
    contentRef
  };
}
const Navigation = forwardRef(
  (props, ref) => {
    const {
      "data-testid": dataTestId,
      style,
      className,
      children,
      ...rest
    } = props;
    const slots = useMemo(() => getSlots(children), [children]);
    const collapsingNav = useCollapsingNav();
    const intl = useIntl();
    return /* @__PURE__ */ jsxs(Fragment, { children: [
      /* @__PURE__ */ jsxs(Menu, { children: [
        /* @__PURE__ */ jsx(Menu.Trigger, { children: /* @__PURE__ */ jsx(
          Button,
          {
            "aria-label": intl.formatMessage({
              defaultMessage: "Open app navigation",
              id: "WCZAUtmjwNCm5iwa",
              description: "Label for the menu trigger to open app-header navigation."
            }),
            style: {
              display: collapsingNav.isOverflowing ? void 0 : "none"
            },
            children: /* @__PURE__ */ jsx(Button.Prefix, { children: /* @__PURE__ */ jsx(MenuIcon, {}) })
          }
        ) }),
        /* @__PURE__ */ jsx(Menu.Content, { children: slots.linksProps.map((linkProps) => {
          const {
            isSelected,
            disabled,
            className: navItemClassName,
            ...rest2
          } = linkProps;
          return /* @__PURE__ */ createElement(
            Menu.Link,
            {
              ...rest2,
              key: linkProps.key,
              className: clsx(
                styles.menuLink,
                {
                  [styles.isSelected]: isSelected,
                  [styles.isDisabled]: disabled
                },
                navItemClassName
              ),
              disabled
            }
          );
        }) })
      ] }),
      /* @__PURE__ */ jsxs(
        Flex,
        {
          ...rest,
          ref,
          gap: 4,
          alignItems: "center",
          flexShrink: 1,
          minWidth: 0,
          style,
          className,
          "data-testid": dataTestId,
          children: [
            slots.logo || /* @__PURE__ */ jsx(Logo, {}),
            /* @__PURE__ */ jsx(
              Flex,
              {
                ref: collapsingNav.containerRef,
                flexGrow: 1,
                minWidth: 0,
                hidden: collapsingNav.isOverflowing,
                style: {
                  visibility: collapsingNav.isOverflowing ? "hidden" : void 0,
                  flexBasis: collapsingNav.isOverflowing ? 0 : void 0
                },
                children: /* @__PURE__ */ jsx(
                  Flex,
                  {
                    as: "nav",
                    gap: 4,
                    className: styles.inlineNav,
                    ref: collapsingNav.contentRef,
                    "aria-label": intl.formatMessage({
                      defaultMessage: "App navigation",
                      id: "oA42meteX3Ww5Zo5"
                    }),
                    children: slots.linksProps.map((linkProps) => {
                      const {
                        isSelected,
                        disabled,
                        as = "a",
                        className: navItemClassName,
                        ...rest2
                      } = linkProps;
                      return /* @__PURE__ */ createElement(
                        Button,
                        {
                          ...rest2,
                          key: linkProps.key,
                          as,
                          className: clsx(
                            styles.button,
                            {
                              [styles.isSelected]: isSelected,
                              [styles.isDisabled]: disabled
                            },
                            navItemClassName
                          ),
                          disabled
                        }
                      );
                    })
                  }
                )
              }
            )
          ]
        }
      )
    ] });
  }
);
Navigation.displayName = "AppHeader.Navigation";
const NavigationItem = forwardRef(
  (props, forwardedRef) => /* @__PURE__ */ jsx("div", { ref: forwardedRef, ...props })
);
NavigationItem.displayName = "AppHeader.NavigationItem";
export {
  Navigation,
  NavigationItem
};
