// @dynatrace/strato-components/esm/filters/segment-selector/contexts/SegmentsCache/SegmentsCache.js
import { QueryClient } from "@tanstack/react-query";
import { getCurrentUserDetails } from "@dynatrace-sdk/app-environment";
import {
  filterSegmentsClient
} from "@dynatrace-sdk/client-filter-segment-management";
const segmentQueryClient = new QueryClient({
  defaultOptions: {
    queries: {
      refetchOnWindowFocus: false,
      refetchOnMount: true,
      retry: false,
      staleTime: Infinity
    }
  }
});
const currentUser = getCurrentUserDetails();
function catchError(error) {
  if (error.name === "403" || error.message === "403") {
    segmentsClient.hasAccessRights = false;
    segmentsClient.queryClient.setDefaultOptions({
      queries: { enabled: false }
    });
  }
}
function getSegmentFromCache(id) {
  const allSegments = segmentsClient.queryClient.getQueryData(["getAllSegments"]);
  const singleSegment = segmentsClient.queryClient.getQueryData([
    "getSegment",
    id
  ]);
  return allSegments?.get(id) ?? singleSegment ?? null;
}
const segmentsClient = {
  /** react-query's queryClient used for segments queries. */
  queryClient: segmentQueryClient,
  /** flag for missing access rights (403) for any segments query. */
  hasAccessRights: void 0
};
function processSegment(segment) {
  const { uid, ...remainingSegment } = segment;
  return {
    segment: { ...remainingSegment, id: uid },
    temporarilyApplied: !(segment.isPublic || segment.owner === currentUser.id)
  };
}
async function getSegment(uid) {
  if (!uid) {
    throw new Error("Segment id is required");
  }
  if (segmentsClient.hasAccessRights === false) {
    throw new Error("403");
  }
  if (segmentsClient.queryClient.isFetching({ queryKey: ["getAllSegments"] }) === 0 && !segmentsClient.queryClient.getQueryState([
    "getAllSegments"
  ])?.isInvalidated) {
    const cachedSegment = getSegmentFromCache(uid);
    if (cachedSegment) {
      return cachedSegment;
    }
  }
  return new Promise((resolve, reject) => {
    filterSegmentsClient.getFilterSegment({
      filterSegmentUid: uid,
      addFields: ["VARIABLES"]
    }).then((storeSegment) => {
      const newCachedSegment = processSegment(storeSegment);
      segmentsClient.hasAccessRights = true;
      resolve(newCachedSegment);
    }).catch((error) => {
      catchError(error);
      reject(error);
      return null;
    });
  });
}
async function getAllSegments() {
  if (segmentsClient.hasAccessRights === false) {
    throw new Error("403");
  }
  return new Promise((resolve, reject) => {
    filterSegmentsClient.getFilterSegments({
      addFields: ["VARIABLES"]
    }).then((segments) => {
      const updatedCache = /* @__PURE__ */ new Map();
      segments.filterSegments.forEach((storeSegment) => {
        const newCachedSegment = processSegment(storeSegment);
        updatedCache.set(storeSegment.uid, newCachedSegment);
        segmentsClient.hasAccessRights = true;
      });
      resolve(updatedCache);
    }).catch((error) => {
      catchError(error);
      reject(error);
    });
  });
}
function useSegmentsCache() {
  return {
    getAllSegments,
    getSegment
  };
}
export {
  getSegment,
  segmentsClient,
  useSegmentsCache
};
