// @dynatrace/strato-components/content/health-indicator/HealthIndicator.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var HealthIndicator_exports = {};
__export(HealthIndicator_exports, {
  HealthIndicator: () => HealthIndicator
});
module.exports = __toCommonJS(HealthIndicator_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_lodash_es = require("lodash");
var import_react = require("react");
var import_Label = require("./Label.js");
var import_StatusVisual = require("./StatusVisual.js");
var styles = __toESM(require("./styles.sty.js"));
var import_Visual = require("./Visual.js");
var import_useAriaLabelingProps = require("../../core/hooks/useAriaLabelingProps.js");
var import_useBehavioralTrackingProps = require("../../core/hooks/useBehavioralTrackingProps.js");
var import_useId = require("../../core/hooks/useId.js");
var import_useSafeForwardProps = require("../../core/hooks/useSafeForwardProps.js");
var import_useSlots = require("../../core/hooks/useSlots.js");
var import_Container = require("../../layouts/container/Container.js");
var import_Flex = require("../../layouts/flex/Flex.js");
const COMPONENT_NAME = "HealthIndicator";
const HealthIndicatorRoot = (0, import_react.forwardRef)(
  (props, forwardedRef) => {
    const {
      children,
      id,
      status = "neutral",
      visual = "shape",
      className: consumerClassName,
      style: consumerStyle,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      "data-testid": dataTestId,
      ...remainingProps
    } = props;
    const containerColoring = (0, import_Container.useContainerColoring)();
    const onAccent = containerColoring.variant === "accent";
    const [behavioralTrackingProps, restProps] = (0, import_useBehavioralTrackingProps.useBehavioralTrackingProps)(remainingProps);
    const ariaLabelingProps = (0, import_useAriaLabelingProps.useAriaLabelingProps)(restProps);
    const forwardProps = (0, import_useSafeForwardProps.useSafeForwardProps)(restProps, ariaLabelingProps);
    const {
      "aria-label": ariaLabel,
      "aria-labelledby": ariaLabelledBy,
      ...remainingAriaLabelingProps
    } = ariaLabelingProps;
    const labelId = (0, import_useId.useId)();
    const [slots] = (0, import_useSlots.useSlots)(children, {
      visual: import_Visual.Visual,
      label: import_Label.Label
    });
    const visualContainerHeight = visual === "icon" ? 10 : 6;
    const hasLabel = (ariaLabelledBy ?? ariaLabel ?? slots.label) !== void 0;
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      import_Flex.Flex,
      {
        id,
        ref: forwardedRef,
        alignItems: "flex-start",
        gap: visual === "icon" ? 8 : 4,
        className: consumerClassName,
        style: {
          minWidth: 0,
          ...consumerStyle
        },
        "data-testid": dataTestId,
        "data-dtrum-mask": dataDtrumMask,
        "data-dtrum-allow": dataDtrumAllow,
        ...behavioralTrackingProps,
        ...forwardProps,
        "data-dt-component": COMPONENT_NAME,
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "div",
            {
              className: styles.visual({ status, onAccent, visualType: visual }),
              role: hasLabel ? "img" : void 0,
              "aria-labelledby": ariaLabelledBy ?? (slots.label && !ariaLabel ? labelId : void 0),
              "aria-label": ariaLabel,
              ...remainingAriaLabelingProps,
              children: !(0, import_lodash_es.isNil)(slots.visual) ? slots.visual : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_StatusVisual.StatusVisual, { status, visual })
            }
          ),
          !(0, import_lodash_es.isNil)(slots.label) && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "div",
            {
              id: labelId,
              style: {
                // Calculate padding-top: (container height for icon (20px) or shape (12px) - 1 line height) / 2
                paddingTop: `calc(-0.5lh + ${visualContainerHeight}px)`,
                minWidth: "0"
              },
              children: slots.label
            }
          )
        ]
      }
    );
  }
);
HealthIndicatorRoot.displayName = COMPONENT_NAME;
const HealthIndicator = Object.assign(HealthIndicatorRoot, {
  Visual: import_Visual.Visual,
  Label: import_Label.Label
});
