// @dynatrace/strato-components/esm/content/expandable-text/ExpandableText.js
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import { forwardRef } from "react";
import { FormattedMessage } from "react-intl";
import {
  expandableTextCSS,
  expandableTextButtonCSS
} from "./ExpandableText.sty.js";
import { useAriaLabelingProps } from "../../core/hooks/useAriaLabelingProps.js";
import { useBehavioralTrackingProps } from "../../core/hooks/useBehavioralTrackingProps.js";
import { useControllableState } from "../../core/hooks/useControllableState.js";
import { useId } from "../../core/hooks/useId.js";
import { useSafeForwardProps } from "../../core/hooks/useSafeForwardProps.js";
const COMPONENT_NAME = "ExpandableText";
const ExpandableText = forwardRef(
  (props, forwardedRef) => {
    const {
      children,
      expandLabel = /* @__PURE__ */ jsx(
        FormattedMessage,
        {
          description: "Shown for expandable text.",
          defaultMessage: "Show more",
          id: "DV3d5lBJda+EHJja"
        }
      ),
      collapseLabel = /* @__PURE__ */ jsx(
        FormattedMessage,
        {
          description: "Shown for collapsible text.",
          defaultMessage: "Show less",
          id: "986buSbL34GVFlBf"
        }
      ),
      defaultExpanded = false,
      expanded,
      onExpandChange,
      className: consumerClassName,
      style: consumerStyle,
      "data-testid": dataTestId,
      "data-dtrum-mask": dataDtrumMask,
      "data-dtrum-allow": dataDtrumAllow,
      ...remainingProps
    } = props;
    const [behavioralTrackingProps, restProps] = useBehavioralTrackingProps(remainingProps);
    const ariaLabelingProps = useAriaLabelingProps(restProps);
    const forwardProps = useSafeForwardProps(restProps, ariaLabelingProps);
    const [isExpanded, setExpanded] = useControllableState({
      defaultValue: defaultExpanded,
      value: expanded,
      onChange: onExpandChange
    });
    const textId = useId();
    const showMoreLessDataDtProperties = `state:${isExpanded ? "less" : "more"}`;
    const beginsWithWhitespace = Array.isArray(children) === false && typeof children === "string" && /^\s/.test(children);
    return /* @__PURE__ */ jsxs(Fragment, { children: [
      /* @__PURE__ */ jsxs(
        "span",
        {
          ref: forwardedRef,
          id: textId,
          role: "region",
          className: clsx(
            expandableTextCSS({ expanded: isExpanded }),
            consumerClassName
          ),
          style: consumerStyle,
          "data-testid": dataTestId,
          "data-dtrum-mask": dataDtrumMask,
          "data-dtrum-allow": dataDtrumAllow,
          ...ariaLabelingProps,
          ...behavioralTrackingProps,
          ...forwardProps,
          "data-dt-component": COMPONENT_NAME,
          children: [
            isExpanded && beginsWithWhitespace && " ",
            isExpanded && children
          ]
        }
      ),
      " ",
      /* @__PURE__ */ jsx(
        "button",
        {
          onClick: () => setExpanded(!isExpanded),
          "aria-controls": textId,
          "aria-expanded": isExpanded,
          className: expandableTextButtonCSS,
          "data-dtrum-mask": dataDtrumMask,
          "data-dtrum-allow": dataDtrumAllow,
          "data-dt-name": "ShowMoreLess",
          "data-dt-properties": showMoreLessDataDtProperties,
          children: isExpanded ? collapseLabel : expandLabel
        }
      )
    ] });
  }
);
ExpandableText.displayName = COMPONENT_NAME;
export {
  ExpandableText
};
