// @dynatrace/strato-components/esm/tables/DataTable/features/SubRows/SubRows.js
import { jsx } from "react/jsx-runtime";
import Spacings from "@dynatrace/strato-design-tokens/spacings";
import { calculateMaxVisibleDepth, isRowDisabled } from "./sub-rows-utils.js";
import { useControllableState } from "../../../../core/hooks/useControllableState.js";
import {
  DATA_TABLE_SUB_ROW_EXPANDER_ICON_SIZE as SUB_ROW_EXPANDER_ICON_SIZE,
  DATA_TABLE_SUB_ROW_INDENT_GRID as SUB_ROW_INDENT_GRID,
  DATA_TABLE_SUB_ROW_DENSITY_PADDING_MAP as SUB_ROW_DENSITY_PADDING_MAP
} from "../../constants.js";
import { useDataTable } from "../../contexts/DataTableContext.js";
import { makeAccessorFunction } from "../../hooks/makeAccessorFunction.js";
import {
  assignRowMethods,
  getMemoOptions
} from "../../hooks/useTable/utils.js";
import { memo } from "../../tanstack-table-core-override/src/utils.js";
function defaultSubRowsAccessor(originalRow) {
  return originalRow?.subRows;
}
const DataTableSubRows = {
  createTable(table) {
    const disabledSubRowsCache = /* @__PURE__ */ new Map();
    const rowLineConfig = /* @__PURE__ */ new Map();
    table.getIsAllRowsExpanded = memo(
      () => [table.getState().expanded, table.getRowModel().flatRows],
      (expandedState, flatRows) => {
        if (typeof expandedState === "boolean") {
          return expandedState === true;
        }
        if (!Object.keys(expandedState).length) {
          return false;
        }
        if (
          // The original code from @tanstack/react-table does not check if a row is even
          // expandable before it counts it as isExpanded.
          flatRows.some((row) => {
            return row.getCanExpand() && !row.getIsExpanded() && !row.getCanExpandDisabled() && !table.getIsAncestorDisabled(row);
          })
        ) {
          return false;
        }
        return true;
      },
      getMemoOptions(table.options, "debugTable", "table.getIsAllRowsExpanded")
    );
    table.toggleAllRowsExpanded = () => {
      if (!table.getIsAllRowsExpanded()) {
        const newExpandedState = table.getExpandedRowModel().flatRows.reduce((expanded, row) => {
          if (row.getCanExpand() && !row.getCanExpandDisabled()) {
            expanded[row.id] = true;
          }
          return expanded;
        }, {});
        table.setExpanded(newExpandedState);
      } else {
        table.setExpanded({});
      }
    };
    table.getSubRowIndicatorColumnId = memo(
      () => [
        table.options.enableExpanding,
        table.options.subRowColumnId,
        table.getState().columnVisibility,
        table.options.columns
      ],
      (enableSubRows, subrowColumnId) => {
        if (enableSubRows !== true) {
          return null;
        }
        if (subrowColumnId) {
          table.remeasureColumn(subrowColumnId);
          return subrowColumnId;
        }
        const visibleLeafColumns = table.getVisibleLeafColumns();
        for (const column of visibleLeafColumns) {
          const columnDef = column.columnDef;
          if (columnDef.meta && columnDef.meta.builtin !== true) {
            table.remeasureColumn(column.id);
            return column.id;
          }
        }
        return null;
      },
      getMemoOptions(
        table.options,
        "debugTable",
        "table.getSubRowIndicatorColumnId"
      )
    );
    table.getSubRowOpenDepth = memo(
      () => [table.getState().expanded],
      (expandedState) => {
        const expandedRowIds = [];
        if (expandedState === true) {
          const rows = table.getExpandedRowModel().rows;
          for (const row of rows) {
            if (row.subRows.length > 0) {
              expandedRowIds.push(row.id);
            }
          }
        } else {
          for (const rowId in expandedState) {
            if (expandedState[rowId]) {
              expandedRowIds.push(rowId);
            }
          }
        }
        return calculateMaxVisibleDepth(
          expandedRowIds,
          table.options.rowIdSeparator
        );
      },
      getMemoOptions(table.options, "debugTable", "table.getSubRowOpenDepth")
    );
    table.getIsAncestorDisabled = memo(
      (row) => [row, table.getPreExpandedRowModel().flatRows],
      (row) => isRowDisabled(disabledSubRowsCache, row),
      getMemoOptions(
        table.options,
        "debugTable",
        "table.getIsAncestorDisabled"
      )
    );
    table.getHasSubRows = () => table.getCanSomeRowsExpand() || table.getIsAllRowsExpanded();
    table.getRowLineConfig = memo(
      (rowId) => [rowId],
      (rowId) => {
        if (!rowId) {
          return [];
        }
        const lineConfig = [];
        const row = table.getRow(rowId);
        for (let i = 1; i <= row.depth; i += 1) {
          if (i === row.depth) {
            const isLastExpandedRowOnLevel = row.getIsLastExpandedOnLevel();
            if (isLastExpandedRowOnLevel) {
              lineConfig.push("LShape");
              continue;
            }
            lineConfig.push("TShape");
            continue;
          }
          lineConfig.push("Line");
        }
        const parentId = row.parentId;
        if (parentId) {
          const parentShapes = rowLineConfig.get(parentId) ?? table.getRowLineConfig(parentId);
          if (parentShapes.at(-1) === "LShape") {
            lineConfig[lineConfig.length - 2] = "Empty";
          }
          for (const [index, shape] of parentShapes.entries()) {
            if (shape === "Empty") {
              lineConfig[index] = "Empty";
            }
          }
        }
        rowLineConfig.set(rowId, lineConfig);
        return lineConfig;
      },
      getMemoOptions(table.options, "debugTable", "table.getRowLineConfig")
    );
    assignRowMethods(table, {
      toggleExpanded(expanded) {
        table.setExpanded((prev) => {
          const exists = prev === true ? true : !!prev?.[this.id];
          let prevExpandedState = {};
          if (prev === true) {
            table.getCoreRowModel().flatRows.forEach((row) => {
              if (!row.getCanExpandDisabled() || !table.getIsAncestorDisabled(row)) {
                prevExpandedState[row.id] = true;
              }
            });
          } else {
            prevExpandedState = prev;
          }
          const shouldExpand = expanded ?? !exists;
          if (!exists && shouldExpand) {
            return {
              ...prevExpandedState,
              [this.id]: true
            };
          }
          if (exists && !shouldExpand) {
            for (const key in prevExpandedState) {
              if (key.startsWith(this.id)) {
                delete prevExpandedState[key];
              }
            }
            const { [this.id]: _, ...rest } = prevExpandedState;
            return rest;
          }
          return prev;
        });
      },
      getIsLastExpandedOnLevel: memo(
        function() {
          return [
            table.getRowOrderRowModel().flatRows,
            table.getState().expanded,
            this.id
          ];
        },
        function(orderedFlatRows) {
          const parentRow = this.getParentRow();
          const orderedSiblings = orderedFlatRows.find(
            (row) => row.id === parentRow?.id
          )?.subRows;
          const lastSibling = orderedSiblings?.at(-1)?.id;
          return lastSibling === this.id;
        },
        getMemoOptions(
          table.options,
          "debugRows",
          "row.getIsLastInExpandedOnLevel"
        )
      ),
      getLastExpandedSubRow: memo(
        function() {
          const state = table.getState();
          return [
            state.rowOrder,
            state.sorting,
            state.expanded,
            table.options.data,
            this.id
          ];
        },
        function() {
          const orderedRow = table.getRowOrderRowModel().flatRows.find((sortedRow) => sortedRow.id === this.id);
          return orderedRow?.getLeafRows().findLast(
            (subRow) => subRow.getParentRow()?.getIsExpanded()
          );
        },
        getMemoOptions(table.options, "debugRows", "row.getLastExpandedSubRow")
      ),
      getIsExpanded() {
        const expanded = table.getState().expanded;
        return !!((expanded === true || expanded[this.id]) && !table.getIsAncestorDisabled(this));
      },
      /** Determines whether the given subRow expander is disabled. */
      getCanExpandDisabled() {
        return table.options.getRowExpandDisabled?.(this.original) ?? false;
      },
      getSubRowDepth() {
        return this.id.split(table.options.rowIdSeparator || ".").length - 1;
      }
    });
  }
};
function getSubRowCellWidth(rowDensity, indentWidth) {
  const rowDensityPadding = SUB_ROW_DENSITY_PADDING_MAP[rowDensity];
  return `calc(${indentWidth + SUB_ROW_EXPANDER_ICON_SIZE + rowDensityPadding}px + ${Spacings.Size4})`;
}
function SubRowsHeaderWidthCorrection(props) {
  const { subRowOpenDepth } = props;
  const table = useDataTable();
  const hasMultiplePages = table.options.enablePagination === true && table.getPageCount() > 1;
  let width;
  if (hasMultiplePages || table.getHasVirtualizedRows()) {
    const indentWidth = SUB_ROW_INDENT_GRID * subRowOpenDepth;
    const rowDensity = table.getVisualVariantOptions().rowDensity;
    width = getSubRowCellWidth(rowDensity, indentWidth);
  }
  return /* @__PURE__ */ jsx(
    "div",
    {
      style: {
        // Expand to fill available space in the cell.
        // This is especially important for right-aligned columns
        // to push the subrow expander all the way to the left.
        flex: 1,
        // Ensure correct width based on the current sub-row max depth,
        // if table has virtualized or paginated rows.
        width
      }
    }
  );
}
function useSubRows(props, options) {
  const { subRows } = props;
  const [openSubRows, setOpenSubRows] = useControllableState({
    value: props.openSubRows,
    defaultValue: props.defaultOpenSubRows,
    onChange: (value) => {
      if (value !== true) {
        props.onOpenSubRowsChange?.(value);
      }
    }
  });
  if (subRows === void 0 || subRows === false) {
    options.enableExpanding = false;
    return;
  }
  options.enableExpanding = true;
  if (typeof subRows === "boolean") {
    options.getSubRows = defaultSubRowsAccessor;
  } else {
    const { accessor, subRowColumnId: subrowColumnId, disableSubRow } = subRows;
    if (accessor === void 0) {
      options.getSubRows = defaultSubRowsAccessor;
    } else if (typeof accessor === "string") {
      if (accessor.includes(".") || accessor.includes("[")) {
        const pathAccessFunction = makeAccessorFunction(accessor);
        options.getSubRows = (originalRow) => pathAccessFunction(originalRow);
      } else {
        options.getSubRows = (originalRow) => originalRow[accessor];
      }
    } else if (typeof accessor === "function") {
      options.getSubRows = (originalRow) => accessor(originalRow);
    }
    if (subrowColumnId) {
      options.subRowColumnId = subrowColumnId;
    }
    if (disableSubRow) {
      options.getRowExpandDisabled = disableSubRow;
    }
  }
  options.state = {
    ...options.state,
    expanded: openSubRows ?? {}
  };
  options.onExpandedChange = setOpenSubRows;
}
export {
  DataTableSubRows,
  SubRowsHeaderWidthCorrection,
  defaultSubRowsAccessor,
  getSubRowCellWidth,
  useSubRows
};
