// @dynatrace/strato-components/esm/tables/DataTable/features/SubRows/SubRowExpanderPath.js
import { jsx } from "react/jsx-runtime";
import clsx from "clsx";
import { memo } from "react";
import * as styles from "./SubRows.sty.js";
import {
  DATA_TABLE_SUB_ROW_DENSITY_PADDING_MAP,
  DATA_TABLE_SUB_ROW_EXPANDER_ICON_SIZE,
  DATA_TABLE_SUB_ROW_SPACER
} from "../../constants.js";
import { useDataTable } from "../../contexts/DataTableContext.js";
function generatePathFromConfig(config, height, verticalAlignment, estimatedRowHeight, rowDensity) {
  const spacer = DATA_TABLE_SUB_ROW_SPACER;
  const rowDensityPadding = DATA_TABLE_SUB_ROW_DENSITY_PADDING_MAP[rowDensity];
  let pathString = "";
  for (const [index, shape] of config.entries()) {
    const oneBaseIndex = index + 1;
    const x = oneBaseIndex * spacer - spacer / 2;
    let yLShape = height / 2;
    if (height > estimatedRowHeight) {
      const expanderSpace = rowDensityPadding + DATA_TABLE_SUB_ROW_EXPANDER_ICON_SIZE / 2;
      if (verticalAlignment === "top") {
        yLShape = expanderSpace;
      } else if (verticalAlignment === "bottom") {
        yLShape = height - expanderSpace;
      }
    }
    const yTShape = -height + yLShape;
    if (shape === "Line") {
      pathString += `M ${x},0 V ${height} `;
    } else if (shape === "LShape") {
      pathString += `M ${x},0 V ${yLShape} h 200 `;
    } else if (shape === "TShape") {
      pathString += `M ${x},0 V ${height} m 0,${yTShape} h 25`;
    }
  }
  return pathString;
}
const InternalMemoizedSubRowExpanderPath = (props) => {
  const { rowId, width, rowLineConfig, rowCanExpand } = props;
  const table = useDataTable();
  const estimatedRowHeight = table.estimateRowHeight({ index: 0 });
  const rowHeight = table.getRowHeight(rowId);
  const visualVariantConfig = table.getVisualVariantOptions();
  const rowDensity = visualVariantConfig.rowDensity;
  const svgWidth = rowCanExpand ? width : width + DATA_TABLE_SUB_ROW_EXPANDER_ICON_SIZE;
  const verticalAlignment = table.getBodyVerticalAlignment();
  return /* @__PURE__ */ jsx(
    "svg",
    {
      className: clsx(
        styles.expandIndentSvg,
        styles.expandIndentSvgSize({ rowCanExpand })
      ),
      style: {
        width: svgWidth
      },
      preserveAspectRatio: "none",
      overflow: "visible",
      viewBox: `0 0 ${svgWidth} ${rowHeight}`,
      children: /* @__PURE__ */ jsx(
        "path",
        {
          d: generatePathFromConfig(
            rowLineConfig,
            rowHeight,
            verticalAlignment,
            estimatedRowHeight,
            rowDensity
          ),
          fill: "transparent",
          vectorEffect: "non-scaling-stroke",
          strokeWidth: 1
        }
      )
    }
  );
};
const SubRowExpanderPath = memo(InternalMemoizedSubRowExpanderPath);
export {
  SubRowExpanderPath
};
