// @dynatrace/strato-components/tables/DataTable/features/LoadingState/LoadingState.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var LoadingState_exports = {};
__export(LoadingState_exports, {
  LoadingState: () => LoadingState,
  PROGRESS_CIRCLE_HEIGHT_BREAKPOINT: () => PROGRESS_CIRCLE_HEIGHT_BREAKPOINT,
  useLoadingState: () => useLoadingState
});
module.exports = __toCommonJS(LoadingState_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_dynamic = require("@vanilla-extract/dynamic");
var import_react = require("react");
var import_react_intl = require("react-intl");
var styles = __toESM(require("./LoadingState.sty.js"));
var import_ProgressCircle = require("../../../../content/progress/ProgressCircle.js");
var import_Flex = require("../../../../layouts/flex/Flex.js");
var import_messages = require("../../messages.js");
const PROGRESS_CIRCLE_HEIGHT_BREAKPOINT = 120;
function useLoadingState(props, options) {
  options.state = { ...options.state, loading: props.loading };
}
function LoadingState(props) {
  const { headerHeight } = props;
  const loadingContainerRef = (0, import_react.useRef)(null);
  const [height, setHeight] = (0, import_react.useState)(void 0);
  const intl = (0, import_react_intl.useIntl)();
  (0, import_react.useEffect)(() => {
    if (!loadingContainerRef.current) {
      return;
    }
    const resizeObserver = new ResizeObserver((entries) => {
      if (entries.length > 0) {
        setHeight(entries[0].contentRect.height);
      }
    });
    resizeObserver.observe(loadingContainerRef.current);
    return () => {
      resizeObserver.disconnect();
    };
  }, []);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_Flex.Flex,
    {
      ref: loadingContainerRef,
      justifyContent: "center",
      alignItems: "center",
      className: styles.loadingState,
      style: (0, import_dynamic.assignInlineVars)({
        [styles.dataTableHeaderHeight]: `${headerHeight}px`
      }),
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_ProgressCircle.ProgressCircle,
        {
          size: height !== void 0 && height > PROGRESS_CIRCLE_HEIGHT_BREAKPOINT ? "large" : "small",
          "aria-label": intl.formatMessage(import_messages.dataTableMessages.loadingData)
        }
      )
    }
  );
}
