// @dynatrace/strato-components/esm/charts/core/components/legend/range/components/RangeLegendBoundaryLabel/RangeLegendBoundaryLabel.js
import { jsx } from "react/jsx-runtime";
import { boundaryLabel } from "./RangeLegendBoundaryLabel.sty.js";
import { ORIENTATION } from "../../../../../constants.js";
import { charsFontWidth12pixelsSize500Map } from "../../../../text-ellipsis/chars-font-width-map.js";
import {
  calculateWidth,
  TEXT_HEIGHT
} from "../../../../text-ellipsis/truncate-string.js";
import {
  DEFAULT_LABEL_FONT_COLOR,
  DEFAULT_LABEL_FONT_SIZE,
  LABEL_FONT_WEIGHT,
  LABEL_GAP
} from "../../constants.js";
import { useOrientation } from "../../context/RangeLegendOrientationContext.js";
import { useRangeLegendHoveredMarkerId } from "../../hooks/store/selectors.js";
import { getHorizontalLabelPosition } from "../../utils/get-horizontal-label-position.js";
import { getVerticalLabelPosition } from "../../utils/get-vertical-label-position.js";
const RangeLegendBoundaryLabel = (props) => {
  const { alignment, endPosition, startPosition, text, ...textProps } = props;
  const { orientation } = useOrientation();
  const hoveredMarkerId = useRangeLegendHoveredMarkerId();
  const labelTextWidth = calculateWidth(text, charsFontWidth12pixelsSize500Map);
  const labelTextHeight = TEXT_HEIGHT;
  const isVertical = orientation === ORIENTATION.Vertical;
  const { x: labelX, y: labelY } = !isVertical ? getHorizontalLabelPosition(
    labelTextWidth,
    startPosition,
    endPosition,
    alignment
  ) : getVerticalLabelPosition(
    labelTextHeight,
    startPosition,
    endPosition,
    alignment
  );
  const isMarkerShown = hoveredMarkerId !== null;
  return /* @__PURE__ */ jsx(
    "text",
    {
      x: labelX + (isVertical ? LABEL_GAP : 0),
      y: labelY + (!isVertical ? LABEL_GAP : 0),
      fontSize: DEFAULT_LABEL_FONT_SIZE,
      fontWeight: LABEL_FONT_WEIGHT,
      fill: DEFAULT_LABEL_FONT_COLOR,
      className: boundaryLabel({ isMarkerShown }),
      role: "listitem",
      "aria-label": `Boundary ${text}`,
      ...textProps,
      children: text
    }
  );
};
export {
  RangeLegendBoundaryLabel
};
