// @dynatrace/strato-components/filters/filter-field/utils/conversion/lezer-tree-conversion/key/key.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var key_exports = {};
__export(key_exports, {
  convertFilterKey: () => convertFilterKey
});
module.exports = __toCommonJS(key_exports);
var import_json_path = require("./json-path.js");
var import_is_node_environment = require("../../../../../../core/utils/is-node-environment.js");
var import_decoration_marks = require("../../../../decorations/decoration-marks.js");
var import_get_text = require("../../../get-text.js");
var import_generic_node = require("../generic-node.js");
var import_node_mutations = require("../node-mutations.js");
var import_identifier = require("../value/identifier.js");
function convertFilterKey(node, doc, parserConfig) {
  const { from, to } = node;
  const cursor = node.node.cursor();
  let convertedNode = {
    type: "Key",
    range: {
      from,
      to
    },
    textValue: "",
    value: ""
  };
  const errors = [];
  let prevNode;
  cursor.iterate((childNode) => {
    const { from: childFrom, to: childTo } = childNode;
    const nodeName = childNode.name;
    switch (nodeName) {
      case "FilterKey": {
        break;
      }
      case "Identifier":
      case "String":
      case "IncompleteString":
      case "Variable": {
        if (prevNode === "JsonPathSquareOpeningBracket") {
          break;
        }
        const {
          convertedNode: { textValue, value },
          isValid: isValidIdentifier,
          positionOperators: { endsWithOperator, startsWithOperator }
        } = (0, import_identifier.convertIdentifier)(childNode, doc, nodeName);
        (0, import_node_mutations.updateNodeValues)(convertedNode, textValue, value);
        if (!isValidIdentifier || nodeName === "IncompleteString" || endsWithOperator || startsWithOperator) {
          (0, import_node_mutations.convertToErrorNode)(convertedNode);
          errors.push({ range: { from, to }, message: "invalidKey" });
        }
        if (endsWithOperator) {
          errors.push({
            range: endsWithOperator,
            message: "unsupportedPositionOperator"
          });
        }
        if (startsWithOperator) {
          errors.push({
            range: startsWithOperator,
            message: "unsupportedPositionOperator"
          });
        }
        break;
      }
      case "InvalidIdentifier": {
        convertedNode = (0, import_generic_node.convertNode)(childNode, doc).convertedNode;
        errors.push({
          range: { from: childFrom, to: childTo },
          message: "invalidKey"
        });
        break;
      }
      case "PositionOperator":
      case "InvalidPositionOperator": {
        const text = (0, import_get_text.getTextFromRange)(childNode, doc);
        (0, import_node_mutations.updateNodeValues)(convertedNode, text);
        (0, import_node_mutations.convertToErrorNode)(convertedNode);
        errors.push({
          range: { from: childFrom, to: childTo },
          message: "unsupportedPositionOperator"
        });
        break;
      }
      case "JsonPath": {
        const { convertedNode: jsonPathNode, errors: jsonPathErrors } = (0, import_json_path.convertJsonPath)(childNode, doc, parserConfig);
        convertedNode = jsonPathNode;
        errors.push(...jsonPathErrors);
        cursor.lastChild();
        break;
      }
      case "\u26A0":
        if (prevNode !== "PositionOperator" && prevNode !== "InvalidPositionOperator") {
          if (prevNode !== "Identifier") {
            errors.push({
              range: { from: childFrom, to: childTo + 1 },
              message: "missingKeyError",
              decoration: import_decoration_marks.preSpotErrorMark
            });
          } else {
            const textValue = (0, import_get_text.getTextFromRange)(childNode, doc);
            if (textValue === ")" || textValue === "]") {
              (0, import_node_mutations.updateNodeValues)(convertedNode, textValue);
              (0, import_node_mutations.convertToErrorNode)(convertedNode);
              errors.push({
                range: { from, to: childTo },
                message: "invalidKey"
              });
            }
          }
        }
        break;
      case "JsonPathSquareOpeningBracket":
        break;
      default:
        if ((0, import_is_node_environment.isDev)()) {
          console.log(
            `${doc.toString()} | Syntax node '${childNode.name}' is not yet handled in the key conversion. Please make sure to handle all nodes accordingly.`
          );
        }
    }
    prevNode = nodeName;
  });
  if (convertedNode.type === "Error") {
    convertedNode.value = convertedNode.textValue;
  }
  return {
    // For an empty error node, we return undefined.
    convertedNode: from === to ? void 0 : convertedNode,
    errors
  };
}
