// @dynatrace/strato-components/editors/base-code-editor/auto-completion/autocompletion.extension.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var autocompletion_extension_exports = {};
__export(autocompletion_extension_exports, {
  autocompletionExtension: () => autocompletionExtension
});
module.exports = __toCommonJS(autocompletion_extension_exports);
var import_autocomplete = require("@codemirror/autocomplete");
var import_render_node = require("../utils/render-node.js");
async function applyAndReopen(view, from, to, completion, apply, reopen) {
  if (apply === void 0 || typeof apply === "string") {
    const value = apply || completion.label;
    view.dispatch({
      changes: { from, to, insert: value },
      selection: { anchor: from + value.length },
      userEvent: "input.complete"
    });
  } else {
    apply(view, completion, from, to);
  }
  if (typeof reopen === "function" ? await reopen(completion) : reopen) {
    (0, import_autocomplete.startCompletion)(view);
  }
}
const autocompletionExtension = (autocompleteProvider, intlProps) => {
  return async (ctx) => {
    const result = await autocompleteProvider(ctx);
    if (!result) {
      return null;
    }
    return {
      ...result,
      options: result.suggestions.map((s, i) => ({
        ...s,
        info: s.popup ? (completion) => (0, import_render_node.render)(completion, s.popup, intlProps) : void 0,
        apply: (view, completion, from, to) => applyAndReopen(
          view,
          from,
          to,
          completion,
          s.apply,
          result.reopen ?? false
        ),
        ...result.keepOrder && { boost: 99 - i }
      }))
    };
  };
};
