// @dynatrace/strato-components/esm/charts/timeseries/layout/TimeseriesHighlightedCanvasLayout.js
import { jsx } from "react/jsx-runtime";
import { TimeseriesGenericCanvasLayout } from "./TimeseriesGenericCanvasLayout.js";
import { applyOpacityToCanvas } from "../../core/utils/apply-opacity-to-canvas.js";
import { safeGet } from "../../core/utils/safe-get.js";
import { useHighlightedSeriesIdState } from "../hooks/store/selectors.js";
import { useTimeseriesFilteredData } from "../hooks/useTimeseriesFilteredData.js";
const TimeseriesHighlightedCanvasLayout = ({
  handleOpacity
}) => {
  const highlightedSeriesId = useHighlightedSeriesIdState();
  const { dataPerVariant, seriesMap } = useTimeseriesFilteredData();
  handleOpacity(applyOpacityToCanvas(highlightedSeriesId));
  if (!highlightedSeriesId) {
    return null;
  }
  const filterByHighlightedId = (series) => series.filter((s) => safeGet(seriesMap, s).id === highlightedSeriesId);
  const filteredDataPerVariant = {
    line: filterByHighlightedId(dataPerVariant.line),
    area: filterByHighlightedId(dataPerVariant.area),
    bar: filterByHighlightedId(dataPerVariant.bar),
    band: filterByHighlightedId(dataPerVariant.band),
    shapePriority: dataPerVariant.shapePriority
  };
  return /* @__PURE__ */ jsx(TimeseriesGenericCanvasLayout, { dataPerVariant: filteredDataPerVariant });
};
export {
  TimeseriesHighlightedCanvasLayout
};
