// @dynatrace/strato-components/esm/charts/timeseries/layout/TimeseriesGenericCanvasLayout.js
import { Fragment, jsx } from "react/jsx-runtime";
import { forwardRef } from "react";
import { Canvas } from "../../core/components/canvas/Canvas.js";
import { CanvasContext } from "../../core/components/canvas/CanvasContext.js";
import { useGridPlotSize } from "../../core/components/chart-grid/hooks/useGridPlotSize.js";
import { OVERLAY_PADDING } from "../../core/components/data-point/constants.js";
import { AreaChartDisplay } from "../components/displays/area/AreaChartDisplay.js";
import { BandChartDisplay } from "../components/displays/band/BandChartDisplay.js";
import { BarChartDisplay } from "../components/displays/bar/BarChartDisplay.js";
import { LineChartDisplay } from "../components/displays/line/LineChartDisplay.js";
import {
  canvasPlot,
  datapointCanvasPlot
} from "../components/overlays/TimeseriesOverlays.sty.js";
import { useXScales } from "../hooks/useXScale.js";
const displays = {
  line: LineChartDisplay,
  bar: BarChartDisplay,
  area: AreaChartDisplay
};
const TimeseriesGenericCanvasLayout = forwardRef(({ dataPerVariant }, canvasRefs) => {
  const { currentScale: xScale } = useXScales();
  const { width, height } = useGridPlotSize();
  const handleRef = (el, index) => {
    if (canvasRefs && typeof canvasRefs !== "function" && canvasRefs.current && el) {
      canvasRefs.current[index] = el;
    }
  };
  return /* @__PURE__ */ jsx(Fragment, { children: dataPerVariant.shapePriority.map((variant, index) => {
    let Display = LineChartDisplay;
    if (variant !== "band") {
      Display = displays[variant];
    }
    const isDottedShape = variant === "line" || variant === "area";
    const canvasContainerClassName = isDottedShape ? datapointCanvasPlot : canvasPlot;
    const canvasWidth = width && isDottedShape ? width + OVERLAY_PADDING : width;
    const canvasHeight = height && isDottedShape ? height + OVERLAY_PADDING : height;
    return /* @__PURE__ */ jsx(
      "div",
      {
        className: canvasContainerClassName,
        children: /* @__PURE__ */ jsx(
          CanvasContext.Provider,
          {
            value: {
              shapeRefs: /* @__PURE__ */ new Map()
            },
            children: /* @__PURE__ */ jsx(
              Canvas,
              {
                ref: (el) => handleRef(el, index),
                width: canvasWidth,
                height: canvasHeight,
                children: variant === "band" ? /* @__PURE__ */ jsx(
                  BandChartDisplay,
                  {
                    series: dataPerVariant.band,
                    xScale
                  }
                ) : /* @__PURE__ */ jsx(
                  Display,
                  {
                    series: dataPerVariant[variant],
                    xScale
                  }
                )
              }
            )
          }
        )
      },
      `highlight-display-${variant}${index}`
    );
  }) });
});
export {
  TimeseriesGenericCanvasLayout
};
